// @universe-design/react/src/date-picker/stories/dateCalendar.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts
import React from 'react';
import type { StoryFn as Story } from '@storybook/react';
import moment from 'moment';
import {
  Input,
  Space,
  Text,
  DateCalendarUI,
  DateCalendar,
  Divider,
  DatePicker,
  DatePickerPanel,
  RangePicker,
  RangePickerPanel,
  RangePickerProvider,
} from '@universe-design/react';

import './todo-calendar.less';
import { isInHoverRange } from '../utils/ui';
import type { DateCalendarUIProps } from '../panels/DateCalendar/interface';
export default {
  title: 'Components/DatePicker/DateCalendar',
  tags: ['autodocs'],
};

export const DateCalendarUIBasic = () => {
  const selectedDates = [moment().add(1, 'day'), moment().add(2, 'day'), moment().add(3, 'day')];

  const selectedRange: [moment.Moment, moment.Moment] = [moment().weekday(1), moment().weekday(3)];

  return (
    <>
      <Divider>show week numbers</Divider>
      <DateCalendarUI selectedDates={selectedDates} hideWeekNumbers={false} />
      <Divider>hide week numbers</Divider>
      <DateCalendarUI selectedDates={selectedDates} />
      <DateCalendarUI selectedRange={selectedRange} />
    </>
  );
};

export const TodoRangeSelection = () => {
  const [range, setRange] = React.useState<[moment.Moment?, moment.Moment?]>([]);
  const phase = React.useRef<'start' | 'end'>('end');
  const [hoverDate, setHoverDate] = React.useState<moment.Moment | null>(null);

  const onDateSelect = (date: moment.Moment) => {
    if (phase.current === 'end') {
      if (range[0] && date.isBefore(range[0])) {
        // 如果新选的结束日期早于开始日期, 则新范围为 [date, date]
        setRange([date, date]);
        return;
      }
      setRange([range[0], date]);
      return;
    }
    if (range[1] && date.isAfter(range[1])) {
      // 如果新选的开始日期晚于结束日期, 则新范围为 [date, date]
      setRange([date, date]);
      return;
    }
    setRange([date, range[1]]);
  };

  const switchToPhase = (_phase: 'start' | 'end') => {
    phase.current = _phase;
  };

  let selectedRange: DateCalendarUIProps['selectedRange'];
  let selectedDates: DateCalendarUIProps['selectedDates'] = [];
  if (range[0] && range[1]) {
    selectedRange = range as DateCalendarUIProps['selectedRange'];
  } else {
    selectedDates = range.filter(x => Boolean(x)) as DateCalendarUIProps['selectedDates'];
  }

  const onDateMouseEnter = (v: moment.Moment) => {
    setHoverDate(v);
  };

  const onDateMouseLeave = () => {
    setHoverDate(null);
  };

  const previewDate = (date: moment.Moment) =>
    Boolean(hoverDate && isInHoverRange(date, 'date', hoverDate, range, phase.current === 'start' ? 0 : 1));

  return (
    <Space align="start">
      <div>
        <Text variant="title1" as="h2">
          Todo 单面板范围选择
        </Text>
        <Text as="p">1. 默认先选结束时间</Text>
        <Text as="p">2. 可以只选结束时间</Text>
        <Text as="p">3. 需要手动点击 input 切换至选开始时间 </Text>
        {/* <Text as="p">4. 需要手动点击 input 切换至选开始时间 </Text> */}
      </div>
      <div>
        <div className="todo-input-fields">
          <Input
            readOnly
            placeholder="选择开始时间"
            value={range[0]?.format('YYYY-MM-DD')}
            onClick={() => switchToPhase('start')}
          />
          <Input
            readOnly
            placeholder="选择结束时间"
            value={range[1]?.format('YYYY-MM-DD')}
            onClick={() => switchToPhase('end')}
          />
        </div>

        <DateCalendarUI
          selectedRange={selectedRange}
          selectedDates={selectedDates}
          onSelect={onDateSelect}
          onDateMouseEnter={onDateMouseEnter}
          onContentMouseLeave={onDateMouseLeave}
          previewDate={previewDate}
        />
      </div>
    </Space>
  );
};

export const DateCalendarBasic = () => (
  <>
    <DateCalendar />
    <DateCalendar hideWeekNumbers={false} />
  </>
);

export const DatePickerBasic = () => (
  <Space align="start">
    <DatePicker hideWeekNumbers={false} />
    <DatePickerPanel hideWeekNumbers={false} />
  </Space>
);

export const RangePickerBasic = () => (
  <Space align="start">
    <RangePicker hideWeekNumbers={false} />
    <RangePickerProvider>
      <RangePickerPanel hideWeekNumbers={false} />
    </RangePickerProvider>
  </Space>
);

export const About: Story = () => (
  <>
    <h1>DateCalendar 组件</h1>
    <p>研发维护人: wanghao.wong</p>
  </>
);
About.storyName = '关于';

export {};
