// @universe-design/react/src/date-picker/stories/homeless.stories.tsx
import React, { useState, useRef } from 'react';
import moment from 'moment';
import 'moment/dist/locale/zh-cn';
import { RangePickerPanel, RangePickerProvider } from '@universe-design/react';
import type { RangePickerState, DatePickerState } from '@universe-design/react';
import { Space, DatePicker, Button, __SCREENSHOT__, Switch, Select, toast, Input, DateCalendar } from '../..';
import PickerPanel from '../panels/picker';
import { DateCalendarUI } from '../panels/DateCalendar/DateCalendarUI';

const { RangePicker } = DatePicker;

export default {
  title: 'Components/DatePicker/临时示例',
};

export const MonthPanelDemo = () => (
  <DateCalendarUI
    selectedDates={[moment('2020-1-1'), moment('2020-04-1')]}
    selectedRange={[moment('2020-07-1'), moment('2025-1-1')]}
    mode="month"
  />
);

export const YearPanelDemo = () => (
  <div>
    <DateCalendarUI mode="year" selectedRange={[moment('2020-1-1'), moment('2025-1-1')]} />
  </div>
);

function ControlRange() {
  const now = moment();
  const [value] = useState<[moment.Moment, moment.Moment]>([now.clone().subtract(3, 'day'), now]);
  return <RangePicker defaultOpen={__SCREENSHOT__} value={value} />;
}

export const RangePickerDemo = () => {
  const now = moment();
  return (
    <div>
      <div>
        <RangePicker
          defaultOpen={__SCREENSHOT__}
          onOpenChange={v => console.log('onOpenChange', v)}
          dateRender={() => 'aaa'}
        />
      </div>
      <div
        style={{
          marginTop: 10,
        }}
      >
        <RangePicker defaultOpen={__SCREENSHOT__} defaultValue={[now.clone().subtract(3, 'day'), now]} />
      </div>
      <div
        style={{
          marginTop: 10,
        }}
      >
        <span>受控模式: </span>
        <ControlRange />
      </div>
    </div>
  );
};

export const Size = () => (
  <>
    <Space>
      <DatePicker defaultOpen={__SCREENSHOT__} size="large" defaultValue={moment()} />
      <DatePicker defaultOpen={__SCREENSHOT__} defaultValue={moment()} />
      <DatePicker defaultOpen={__SCREENSHOT__} size="small" defaultValue={moment()} />
    </Space>
    <Space>
      <DatePicker.RangePicker defaultOpen={__SCREENSHOT__} size="large" defaultValue={[moment(), moment()]} />
      <DatePicker.RangePicker defaultOpen={__SCREENSHOT__} defaultValue={[moment(), moment()]} />
      <DatePicker.RangePicker defaultOpen={__SCREENSHOT__} size="small" defaultValue={[moment(), moment()]} />
    </Space>
  </>
);

export const TriggerElement = () => {
  const btnRef = useRef();
  const [rangeValue, setRangeValue] = React.useState<[moment.Moment?, moment.Moment?]>([undefined, undefined]);

  const renderDatePickerTriggerElement = ({ text, hoverText, value }: DatePickerState) => {
    console.log(value);
    return <Button ref={btnRef}>Data: {text || hoverText}</Button>;
  };

  const renderRangePickerTriggerElement = ({ startText, endText }: RangePickerState) => (
    <Button>
      Data: {startText} - {endText}{' '}
    </Button>
  );

  return (
    <Space>
      <DatePicker defaultOpen={__SCREENSHOT__} triggerElement={renderDatePickerTriggerElement} />
      <DatePicker defaultOpen={__SCREENSHOT__} triggerElement={<Button>123</Button>} />
      <DatePicker.RangePicker
        defaultOpen={__SCREENSHOT__}
        value={rangeValue}
        triggerElement={renderRangePickerTriggerElement}
        onChange={setRangeValue}
      />
      <DatePicker.RangePicker
        value={rangeValue}
        triggerElement={
          <Button>
            {rangeValue[0]?.format('MM/DD/YYYY') ?? '未选择'}
            {' - '}
            {rangeValue[1]?.format('MM/DD/YYYY') ?? '未选择'}
          </Button>
        }
        onChange={setRangeValue}
      />
      <DatePicker.RangePicker defaultOpen={__SCREENSHOT__} triggerElement={<Button>123</Button>} />
      <Button onClick={() => console.log(btnRef.current)}>console ref</Button>
    </Space>
  );
};

export const PanelDemo = () => (
  <>
    <Button onClick={() => moment().locale('zh-cn')}>中文</Button>
    <Button onClick={() => moment().locale('en-us')}>英文</Button>
    <Button onClick={() => console.log(moment())}>打印</Button>
    <PickerPanel mode="date" footer={false} />
  </>
);

export const weekStart = () => (
  <div>
    <div>
      周日开始: <DatePicker defaultOpen={__SCREENSHOT__} startOfWeek={0} />
    </div>
    <div
      style={{
        marginTop: 10,
      }}
    >
      周一开始: <DatePicker defaultOpen={__SCREENSHOT__} startOfWeek={1} />
    </div>
    <div
      style={{
        marginTop: 10,
      }}
    >
      周六开始: <DatePicker defaultOpen={__SCREENSHOT__} startOfWeek={6} />
    </div>
  </div>
);

export const TriggerOnClear = () => (
  <div>
    <DatePicker defaultOpen={__SCREENSHOT__} allowClear onClear={() => console.log('on datePicker clear')} />
    <RangePicker defaultOpen={__SCREENSHOT__} allowClear onClear={() => console.log('on rangePicker clear')} />
  </div>
);

export const ShowTime = () => {
  const [showTime, setShowTime] = useState(false);
  const [needConfirm, setNeedConfirm] = useState(true);
  const [showNow, setShowNow] = useState(showTime);
  const [format, setFormat] = useState('YYYY DD MM HH:mm:ss');
  return (
    <Space>
      <Space direction="vertical">
        <span>
          showTime：
          <Switch onChange={v => setShowTime(v)} checked={showTime} />
        </span>
        <span>
          开启二次确认：
          <Switch checked={needConfirm} onChange={v => setNeedConfirm(v)} />
        </span>
        <span>
          showNow
          <Switch checked={showNow} onChange={v => setShowNow(v)} />
        </span>
        <span>
          时间格式：
          <Select
            value={format}
            onChange={value => setFormat(value as string)}
            options={[
              { value: 'YYYY DD MM HH:mm:ss' },
              { value: 'YYYY DD MM HH:mm' },
              { value: 'YYYY DD MM hh:mm:ss' },
            ]}
          />
        </span>
      </Space>

      <DatePicker
        format={format}
        needConfirm={needConfirm}
        showTime={showTime}
        defaultPickerValue={moment('01/01/2022')}
        showNow={showNow}
        onChange={v => toast.success(`选择了 ${moment(v).format(format)}`)}
        defaultOpen
        open
      />
    </Space>
  );
};

export const defaultPickerValue = () => (
  <>
    <p>defaultPickerValue: 10/10/2022, 10/10/2025</p>
    <RangePicker onChange={v => console.log(v)} defaultPickerValue={[moment('10/10/2022'), moment('10/10/2025')]} />
    <p>defaultPickerValue: 10/10/2022</p>
    <RangePicker onChange={v => console.log(v)} defaultPickerValue={[moment('10/10/2022')] as any} />
    <p>defaultPickerValue: null, 10/10/2022</p>
    <RangePicker onChange={v => console.log(v)} defaultPickerValue={[null, moment('10/10/2022')] as any} />
  </>
);

export const PanelWithInput = () => (
  <RangePicker
    renderExtraFooter={() => (
      <>
        <Input />
        <input type="text" />
        <Select options={[{ value: 1 }]} mode="multiple" />
      </>
    )}
    onChange={v => console.log(v)}
    defaultPickerValue={[null, moment('10/10/2022')] as any}
  />
);

export const Cursor = () => {
  const [customeControlledOpen, setCustomeControlledOpen] = React.useState(true);

  return (
    <>
      <div style={{ marginBottom: 500 }}>
        1. normal
        <RangePicker />
      </div>
      <div style={{ marginBottom: 500 }}>
        2. defaultOpen
        <RangePicker defaultOpen />
      </div>
      <div style={{ marginBottom: 500 }}>
        3. 强制 open
        <RangePicker open />
      </div>
      <div style={{ marginBottom: 500 }}>
        4. 用户可使用 onOpenChange 手动控制 open
        <RangePicker open={customeControlledOpen} onOpenChange={_open => setCustomeControlledOpen(_open)} />
      </div>
    </>
  );
};

export const RenderExtraFooter = () => {
  const [value, setValue] = React.useState<moment.Moment | undefined>(moment());
  const [tempValue, setTempValue] = React.useState<moment.Moment | undefined>(value);
  return (
    <>
      <h6>用户可使用 renderExtraFooter 实现二次确认功能 {value?.format('MM-DD')}</h6>
      <DatePicker
        needConfirm
        onSelect={v => {
          setTempValue(v);
        }}
        onChange={v => {
          setValue(v);
        }}
        value={value}
        renderExtraFooter={ctx => (
          <Space
            style={{
              padding: 8,
            }}
          >
            <Button
              size="small"
              type="text"
              onClick={() => {
                ctx.setValue(undefined, false);
                setTempValue(undefined);
              }}
            >
              清空
            </Button>
            <Button
              size="small"
              onClick={() => {
                ctx.setValue(value, false);
                setTempValue(value);
              }}
            >
              取消
            </Button>
            <Button
              size="small"
              type="primary"
              onClick={() => {
                ctx.setValue(tempValue, true);
                ctx.setOpened(false);
              }}
            >
              确定
            </Button>
          </Space>
        )}
      />
    </>
  );
};

export const Disabled = () => {
  const minDate = moment('2024/01/20');
  const maxDate = moment('2027/01/20');
  const props = {
    maxDate,
    minDate,
  };
  return (
    <>
      <p>
        可选范围： minDate: {minDate.format('YYYY-MM-DD')} maxDate: {maxDate.format('YYYY-MM-DD')}
      </p>
      <DateCalendarUI {...props} />
      <DateCalendar {...props} />
      <DatePicker {...props} />
      <RangePickerProvider {...props}>
        <RangePickerPanel />
      </RangePickerProvider>
      <RangePicker {...props} />
    </>
  );
};
