// @universe-design/react/src/date-picker/stories/screenshot.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C3ReuseDeprecated from '../demo/3.reuse.deprecated';
import CAllKindsOfDateCalendar from '../demo/allKindsOfDateCalendar';
import CCell from '../demo/cell';
import CConfirmButtonCommit from '../demo/confirm-button-commit';
import CDateBodyRangeScreenshot from '../demo/date-body-range.screenshot';
import CDateBody from '../demo/date-body';
import CDateTimePicker from '../demo/dateTimePicker';
import CExtraFooterScreenshot from '../demo/extra-footer.screenshot';
import CExtraFooter from '../demo/extra-footer';
import CFormatDisabledScreenshot from '../demo/format-disabled.screenshot';
import CFormatScreenshot from '../demo/format.screenshot';
import CMinMaxDateScreenshot from '../demo/min-maxDate.screenshot';
import CPickDemo from '../demo/pick-demo';
import CPickerDemoScreenshot from '../demo/picker-demo.screenshot';
import CRangePanelDisabled from '../demo/range-panel-disabled';
import CRangePickerScreenshot from '../demo/range-picker.screenshot';
import CReuseScreenshot from '../demo/reuse.screenshot';
import CShowTimeScreenshot from '../demo/show-time.screenshot';
import CShowWeekNumberScreenshot from '../demo/show-week-number.screenshot';
import CSizeScreenshot from '../demo/size.screenshot';
if (!(C3ReuseDeprecated as Story).storyName) {
  (C3ReuseDeprecated as Story).storyName = '3 reuse deprecated';
}
if (!(CAllKindsOfDateCalendar as Story).storyName) {
  (CAllKindsOfDateCalendar as Story).storyName = 'All kinds of date calendar';
}
if (!(CCell as Story).storyName) {
  (CCell as Story).storyName = 'Cell';
}
if (!(CConfirmButtonCommit as Story).storyName) {
  (CConfirmButtonCommit as Story).storyName = 'Confirm button commit';
}
if (!(CDateBodyRangeScreenshot as Story).storyName) {
  (CDateBodyRangeScreenshot as Story).storyName = 'Date body range screenshot';
}
if (!(CDateBody as Story).storyName) {
  (CDateBody as Story).storyName = 'Date body';
}
if (!(CDateTimePicker as Story).storyName) {
  (CDateTimePicker as Story).storyName = 'Date time picker';
}
if (!(CExtraFooterScreenshot as Story).storyName) {
  (CExtraFooterScreenshot as Story).storyName = 'Extra footer screenshot';
}
if (!(CExtraFooter as Story).storyName) {
  (CExtraFooter as Story).storyName = 'Extra footer';
}
if (!(CFormatDisabledScreenshot as Story).storyName) {
  (CFormatDisabledScreenshot as Story).storyName = 'Format disabled screenshot';
}
if (!(CFormatScreenshot as Story).storyName) {
  (CFormatScreenshot as Story).storyName = 'Format screenshot';
}
if (!(CMinMaxDateScreenshot as Story).storyName) {
  (CMinMaxDateScreenshot as Story).storyName = 'Min max date screenshot';
}
if (!(CPickDemo as Story).storyName) {
  (CPickDemo as Story).storyName = 'Pick demo';
}
if (!(CPickerDemoScreenshot as Story).storyName) {
  (CPickerDemoScreenshot as Story).storyName = 'Picker demo screenshot';
}
if (!(CRangePanelDisabled as Story).storyName) {
  (CRangePanelDisabled as Story).storyName = 'Range panel disabled';
}
if (!(CRangePickerScreenshot as Story).storyName) {
  (CRangePickerScreenshot as Story).storyName = 'Range picker screenshot';
}
if (!(CReuseScreenshot as Story).storyName) {
  (CReuseScreenshot as Story).storyName = 'Reuse screenshot';
}
if (!(CShowTimeScreenshot as Story).storyName) {
  (CShowTimeScreenshot as Story).storyName = 'Show time screenshot';
}
if (!(CShowWeekNumberScreenshot as Story).storyName) {
  (CShowWeekNumberScreenshot as Story).storyName = 'Show week number screenshot';
}
if (!(CSizeScreenshot as Story).storyName) {
  (CSizeScreenshot as Story).storyName = 'Size screenshot';
}

export default {
  title: 'Components/DatePicker/截图测试',
  tags: ['autodocs'],
};

export {
  C3ReuseDeprecated,
  CAllKindsOfDateCalendar,
  CCell,
  CConfirmButtonCommit,
  CDateBodyRangeScreenshot,
  CDateBody,
  CDateTimePicker,
  CExtraFooterScreenshot,
  CExtraFooter,
  CFormatDisabledScreenshot,
  CFormatScreenshot,
  CMinMaxDateScreenshot,
  CPickDemo,
  CPickerDemoScreenshot,
  CRangePanelDisabled,
  CRangePickerScreenshot,
  CReuseScreenshot,
  CShowTimeScreenshot,
  CShowWeekNumberScreenshot,
  CSizeScreenshot,
};
