// @universe-design/react/src/date-picker/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Single from '../demo/0.single';
import C1RangePicker from '../demo/1.range-picker';
import C10DateCalendarUi from '../demo/10.date-calendar-ui';
import C11DateCalendar from '../demo/11.date-calendar';
import C2Format from '../demo/2.format';
import C3Reuse from '../demo/3.reuse';
import C4DateRender from '../demo/4.date-render';
import C5ShowTime from '../demo/5.show-time';
import C6NeedConfirm from '../demo/6.needConfirm';
import C7DefaultPickerValue from '../demo/7.defaultPickerValue';
import C8DateBodyRange from '../demo/8.date-body-range';
import C9UnlinkPanels from '../demo/9.unlink-panels';
import CMinDate from '../demo/min-date';
import CShowWeekNumber from '../demo/show-week-number';
if (!(C0Single as Story).storyName) {
  (C0Single as Story).storyName = '0. 单个选择';
  (C0Single as Story).parameters = {
    docs: {
      description: {
        story: '使用`DatePicker`选择单个日、月、年。\n\n',
      },
    },
  };
}
if (!(C1RangePicker as Story).storyName) {
  (C1RangePicker as Story).storyName = '1. 范围选择';
  (C1RangePicker as Story).parameters = {
    docs: {
      description: {
        story: '使用`RangePicker`进行范围选择，支持日、月、年选择。\n\n',
      },
    },
  };
}
if (!(C10DateCalendarUi as Story).storyName) {
  (C10DateCalendarUi as Story).storyName = '10. 日期面板 UI 组件';
  (C10DateCalendarUi as Story).parameters = {
    docs: {
      description: {
        story: '可以使用 `DateCalendarUI` 组件高度自定义选择逻辑。\n\n',
      },
    },
  };
}
if (!(C11DateCalendar as Story).storyName) {
  (C11DateCalendar as Story).storyName = '11. 日期面板组件';
  (C11DateCalendar as Story).parameters = {
    docs: {
      description: {
        story: '可以使用 `DateCalendar` 只使用面板能力。\n\n',
      },
    },
  };
}
if (!(C2Format as Story).storyName) {
  (C2Format as Story).storyName = '2. 自定义日期显示';
  (C2Format as Story).parameters = {
    docs: {
      description: {
        story:
          '通过`format`参数自定义 [Moment 日期格式](https://momentjs.com/docs/#/displaying/)或任意日期格式。通过`parse`解析自定义的日期格式为 Moment 对象。\n\n',
      },
    },
  };
}
if (!(C3Reuse as Story).storyName) {
  (C3Reuse as Story).storyName = '3. 只使用面板';
  (C3Reuse as Story).parameters = {
    docs: {
      description: {
        story: '只使用选择面板，不显示输入框。\n\n',
      },
    },
  };
}
if (!(C4DateRender as Story).storyName) {
  (C4DateRender as Story).storyName = '4. 给日期单元格增加提示';
  (C4DateRender as Story).parameters = {
    docs: {
      description: {
        story: '可以使用 `dateRender` 属性给禁用日期单元格增加提示\n\n',
      },
    },
  };
}
if (!(C5ShowTime as Story).storyName) {
  (C5ShowTime as Story).storyName = '5. 时间选择';
  (C5ShowTime as Story).parameters = {
    docs: {
      description: {
        story: '可以使用 `showTime` 属性集成时间选择。当 showTime 为对象时，会传递给内部的 TimePicker。\n\n',
      },
    },
  };
}
if (!(C6NeedConfirm as Story).storyName) {
  (C6NeedConfirm as Story).storyName = '6. 二次确认';
  (C6NeedConfirm as Story).parameters = {
    docs: {
      description: {
        story:
          '可以使用 `needConfirm` 属性来定义是否需要二次确认提交。当 needConfirm 为 true 时，会显示确认按钮，选择日期或时间之后，点击确认按钮才会真正提交。当设置了`showTime`时，默认 `needConfirm` 为 true, 其他默认为 false。\n\n',
      },
    },
  };
}
if (!(C7DefaultPickerValue as Story).storyName) {
  (C7DefaultPickerValue as Story).storyName = '7. 设置面板默认显示时间';
  (C7DefaultPickerValue as Story).parameters = {
    docs: {
      description: {
        story: '可以使用 `defaultPickerValue` 属性来设置当 vale 为空时，面板显示的时间。\n\n',
      },
    },
  };
}
if (!(C8DateBodyRange as Story).storyName) {
  (C8DateBodyRange as Story).storyName = '8. 使用范围选择面板';
  (C8DateBodyRange as Story).parameters = {
    docs: {
      description: {
        story: '可以单独使用范围选择面板。\n\n',
      },
    },
  };
}
if (!(C9UnlinkPanels as Story).storyName) {
  (C9UnlinkPanels as Story).storyName = '9. 取消面板联动逻辑';
  (C9UnlinkPanels as Story).parameters = {
    docs: {
      description: {
        story: '可以使用 `panelIntervalLimits` API 取消面板联动。\n\n',
      },
    },
  };
}
if (!(CMinDate as Story).storyName) {
  (CMinDate as Story).storyName = '13. 最大、最小时间';
  (CMinDate as Story).parameters = {
    docs: {
      description: {
        story:
          '可以使用 `minDate`、`maxDate` 限制用户可选时间范围。虽然 `disableDate` API 也能实现此目标，但 `disableDate` 在年面板上会调用非常多次，容易引发性能问题。\n\n',
      },
    },
  };
}
if (!(CShowWeekNumber as Story).storyName) {
  (CShowWeekNumber as Story).storyName = '14. 日期面板显示周数';
  (CShowWeekNumber as Story).parameters = {
    docs: {
      description: {
        story: '可以使用 `hideWeekNumbers` 属性使日期选择面板显示或隐藏周数。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/DatePicker/官网示例',
  tags: ['autodocs'],
};

export {
  C0Single,
  C1RangePicker,
  C10DateCalendarUi,
  C11DateCalendar,
  C2Format,
  C3Reuse,
  C4DateRender,
  C5ShowTime,
  C6NeedConfirm,
  C7DefaultPickerValue,
  C8DateBodyRange,
  C9UnlinkPanels,
  CMinDate,
  CShowWeekNumber,
};
