// @universe-design/react/src/overflow/test-demo/basic.tsx
import * as React from 'react';
import type { OverflowProps } from '@universe-design/react';
import { Overflow } from '@universe-design/react';

export interface OverflowTestDemoBasicProps
  extends Pick<OverflowProps<string>, 'maxCount' | 'restPosition' | 'suffixWrap'> {
  /**
   * 容器的宽度
   */
  containerWidth: number;
  /**
   * suffix 的宽度
   */
  suffixWidth?: number;
  /**
   * items
   */
  items: { key: string; width: number }[];
}

export default (props: OverflowTestDemoBasicProps) => {
  const { containerWidth, suffixWidth, items, ...restProps } = props;
  const height = 20;
  return (
    <div style={{ width: containerWidth }}>
      <Overflow
        {...restProps}
        keys={items.map(o => o.key)}
        itemRenderer={key => (
          <div
            style={{
              width: items.find(o => o.key === key)?.width,
              height,
              backgroundColor: 'gray',
            }}
          >
            {key}
          </div>
        )}
        restRenderer={omittedKeys => (
          <div
            style={{
              height,
              backgroundColor: 'blue',
            }}
          >
            {`+${omittedKeys.length}`}
          </div>
        )}
        suffixRenderer={
          suffixWidth
            ? () => (
                <div
                  style={{
                    width: suffixWidth,
                    height,
                    color: 'red',
                  }}
                />
              )
            : undefined
        }
      />
    </div>
  );
};
