// @universe-design/react/es/shared/components/list/ListItem.js
import _extends from "@babel/runtime/helpers/esm/extends";
import ListCheckBoldOutlined from '@universe-design/icons-react/ListCheckBoldOutlined';
import clsx from 'clsx';
import React from 'react';
import stylex from '@universe-design/stylex';
import { cssTypographyBody2, cssFlexBase, cssTypographyHeadline, cssTypographyBody0 } from '../../../stylex';
import { isNonNullable } from '../../../util';
import { useCssScope } from '../../../config-provider/hooks';
import { TextOverflow } from '../text-overflow';
import './index.less';
const cssListItemLike = stylex.compose(cssFlexBase, cssTypographyBody0);
export const ListItemLike = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    className,
    ...restProps
  } = _ref;
  const cssScope = useCssScope('list__itemLike');
  return /*#__PURE__*/React.createElement("div", _extends({
    ref: ref,
    className: clsx(cssScope(), stylex(cssListItemLike), className)
  }, restProps));
});
export const ListItem = /*#__PURE__*/React.forwardRef((_ref2, ref) => {
  let {
    className,
    disabled = false,
    selected = false,
    active = false,
    before,
    after,
    children,
    showCheckmark = true,
    sx,
    ...restProps
  } = _ref2;
  const cssScope = useCssScope('list__item');
  const itemLikeCssScope = useCssScope('list__itemLike');
  return /*#__PURE__*/React.createElement("li", _extends({
    ref: ref,
    className: clsx(className, cssScope(), itemLikeCssScope(), stylex(cssListItemLike, ...(sx !== null && sx !== void 0 ? sx : [])), disabled && cssScope('-disabled'), selected && cssScope('-selected'), active && cssScope('-active'))
  }, restProps), isNonNullable(before) && /*#__PURE__*/React.createElement("div", {
    className: cssScope('__before')
  }, before), isNonNullable(children) && /*#__PURE__*/React.createElement("div", {
    className: cssScope('__content')
  }, children), isNonNullable(after) ? /*#__PURE__*/React.createElement("div", {
    className: cssScope('__after')
  }, after) : showCheckmark ? /*#__PURE__*/React.createElement(ListCheckBoldOutlined, {
    className: clsx(cssScope('__checkmark'))
  }) : null);
});
export const ListItemTitle = /*#__PURE__*/React.forwardRef((_ref3, ref) => {
  let {
    className,
    children,
    before,
    textOverflow = true,
    ...restProps
  } = _ref3;
  const cssScope = useCssScope('list__item__title');
  return /*#__PURE__*/React.createElement("div", _extends({
    ref: ref,
    className: clsx(cssScope(), className)
  }, restProps), isNonNullable(before) && /*#__PURE__*/React.createElement("div", {
    className: cssScope('__before')
  }, before), isNonNullable(children) && (textOverflow ? /*#__PURE__*/React.createElement(TextOverflow, {
    as: "div",
    lineClamp: 1
  }, children) : children));
});
export const ListItemDesc = /*#__PURE__*/React.forwardRef((_ref4, ref) => {
  let {
    className,
    ...restProps
  } = _ref4;
  const cssScope = useCssScope('list__item__description');
  return /*#__PURE__*/React.createElement("div", _extends({
    ref: ref,
    className: clsx(cssScope(), stylex(cssTypographyBody2), className)
  }, restProps));
});
export const ListDivider = /*#__PURE__*/React.forwardRef((_ref5, ref) => {
  let {
    className,
    fluid = false,
    marginBlock = 'both',
    ...restProps
  } = _ref5;
  const cssScope = useCssScope('list__divider');
  return /*#__PURE__*/React.createElement("hr", _extends({
    ref: ref,
    className: clsx(cssScope(), className, fluid && cssScope('--fluid'), cssScope(`--margin-${marginBlock}`))
  }, restProps));
});
export const ListCaption = /*#__PURE__*/React.forwardRef((_ref6, ref) => {
  let {
    className,
    sx,
    ...restProps
  } = _ref6;
  const cssScope = useCssScope('list__caption');
  const itemLikeCssScope = useCssScope('list__itemLike');
  return /*#__PURE__*/React.createElement("div", _extends({
    ref: ref,
    className: clsx(cssScope(), itemLikeCssScope(), stylex(cssListItemLike, cssTypographyHeadline, ...(sx || [])), className)
  }, restProps));
});
export const ListEmpty = /*#__PURE__*/React.forwardRef((_ref7, ref) => {
  let {
    className,
    ...restProps
  } = _ref7;
  const cssScope = useCssScope('list__empty');
  const itemLikeCssScope = useCssScope('list__itemLike');
  return /*#__PURE__*/React.createElement("div", _extends({
    ref: ref,
    className: clsx(cssScope(), itemLikeCssScope(), stylex(cssListItemLike), className)
  }, restProps));
});
