// @universe-design/react/lib/text/Text.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.fontStyles = exports.Text = void 0;
var _uzf2tb = _interopRequireDefault(require("./../.stylex/uzf2tb7"));
var _u1vjduc = _interopRequireDefault(require("./../.stylex/u1vjduc2"));
var _u1443mj = _interopRequireDefault(require("./../.stylex/u1443mj0"));
var _u7llypb = _interopRequireDefault(require("./../.stylex/u7llypb"));
var _uuojwkd = _interopRequireDefault(require("./../.stylex/uuojwkd"));
var _u1d8k0i = _interopRequireDefault(require("./../.stylex/u1d8k0i0"));
var _u1ust6hi = _interopRequireDefault(require("./../.stylex/u1ust6hi"));
var _uuubajc = _interopRequireDefault(require("./../.stylex/uuubajc"));
var _u8wgnjj = _interopRequireDefault(require("./../.stylex/u8wgnjj"));
var _uitfv9r = _interopRequireDefault(require("./../.stylex/uitfv9r"));
var _u1pdmi = _interopRequireDefault(require("./../.stylex/u1pdmi33"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _react2 = require("@universe-design/stylex/react");
var _stylex2 = require("../stylex");
var _forwardRef = require("../util/forwardRef");
var _useTextOverflowPopperSidecar = require("../config-provider/hooks/useTextOverflowPopperSidecar");
var _dom = require("../util/dom");
var _configProvider = require("../config-provider");
var _util = require("../util");
//.uns1q8e{color:@udtoken-component-text-disabled-loading;}
//.udrzuqr{cursor:not-allowed;}
//.u1xqy1su{user-select:none;}
//.u1j4dglz{font-style:italic;}
//.u1j59e10{text-align:justify;}
//.u12ymhq5{text-align:end;}
//.u17mccla{text-align:center;}
//.upgzoln{text-align:start;}
//.u14t3ns0{display:inline-block;}
//.uhx3js9{word-wrap:break-word;}
//.uokg9q4{overflow-wrap:break-word;}
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const fontStyles = exports.fontStyles = {
  title0: _stylex2.cssTypographyTitle0,
  title1: _stylex2.cssTypographyTitle1,
  title2: _stylex2.cssTypographyTitle2,
  title3: _stylex2.cssTypographyTitle3,
  title4: _stylex2.cssTypographyTitle4,
  title5: _stylex2.cssTypographyTitle5,
  headline: _stylex2.cssTypographyHeadline,
  caption0: _stylex2.cssTypographyCaption0,
  caption1: _stylex2.cssTypographyCaption1,
  caption1Hack: _stylex2.cssTypographyCaption1Hack,
  body0: _stylex2.cssTypographyBody0,
  body2: _stylex2.cssTypographyBody2
};

/**
 * @docgen
 */

const Text = exports.Text = (0, _forwardRef.genericForwardRef)((_ref, ref) => {
  let {
    as: C = 'span',
    variant = 'body0',
    lineClamp,
    disabled = false,
    tooltipProps,
    alignment,
    italic = false,
    ...props
  } = _ref;
  /**
   * ====================
   *   Props
   * ====================
   */
  const {
    children
  } = props;
  const noClamp = !(0, _util.isNonNullable)(lineClamp);
  const oneLineClamp = !noClamp && lineClamp <= 1;
  const multilineClamp = !oneLineClamp && !noClamp;

  /**
   * ====================
   *   Context
   * ====================
   */
  const cssScope = (0, _configProvider.useCssScope)('text');
  const TextOverflowPopper = (0, _useTextOverflowPopperSidecar.useTextOverflowPopperSidecar)();

  /**
   * ====================
   *   Styles
   * ====================
   */
  const cProps = (0, _react2.useMergeComponentProps)(props, {
    className: cssScope(),
    sx: React.useMemo(() => [{
      "props": {
        "onut5h": _uitfv9r.default,
        "Btd35i7": _u1pdmi.default
      }
    }, variant !== 'plain' && fontStyles[variant],
    // 如果 span 并且开了 one line clamp 则需要使用 inline-block
    C === 'span' && oneLineClamp && {
      "props": {
        "mc9l5x": _u8wgnjj.default
      }
    },
    // line clamp 里面有 display: -webkit-box 要放在其他的 display 后面
    (() => {
      if (oneLineClamp) {
        return _stylex2.sxOverflowEllipsis;
      }
      if (multilineClamp) {
        return _stylex2.sxLineClamp;
      }
    })(),
    // 默认从父级继承或使用 initial 值
    alignment && {
      start: {
        "props": {
          "fsow6f": _uuubajc.default
        }
      },
      center: {
        "props": {
          "fsow6f": _u1ust6hi.default
        }
      },
      end: {
        "props": {
          "fsow6f": _u1d8k0i.default
        }
      },
      justify: {
        "props": {
          "fsow6f": _uuojwkd.default
        }
      }
    }[alignment], italic && {
      "props": {
        "B80ckks": _u7llypb.default
      }
    }, disabled && {
      "props": {
        "sj55zd": _uzf2tb.default,
        "Bceei9c": _u1vjduc.default,
        "famaaq": _u1443mj.default
      }
    }], [C, alignment, disabled, italic, multilineClamp, oneLineClamp, variant]),
    style: {
      WebkitLineClamp: multilineClamp ? lineClamp : undefined
    }
  });

  /**
   * ====================
   *   Render
   * ====================
   */
  const node = (0, _react2.jsx)(C, (0, _extends2.default)({
    ref: ref
  }, cProps));
  if (noClamp) {
    return node;
  }
  const checkOverflow = oneLineClamp ? _dom.isHorizontalOverflowed : _dom.isVerticalOverflowed;
  return (0, _react2.jsx)(TextOverflowPopper, (0, _extends2.default)({
    title: children,
    checkOverflow
  }, tooltipProps), node);
});

/**
 * @docgen
 */

if (process.env.NODE_ENV !== "production") {
  Text.displayName = 'Text';
}
