// @universe-design/react/es/table/hooks/useSorter.js
import React, { useCallback, useMemo, useState } from 'react';
import clsx from 'clsx';
import { Tooltip } from '../../tooltip';
import ExpandDownFilled from '@universe-design/icons-react/ExpandDownFilled';
import ExpandUpFilled from '@universe-design/icons-react/ExpandUpFilled';
import SortZToAOutlined from '@universe-design/icons-react/SortZToAOutlined';
import SortAToZOutlined from '@universe-design/icons-react/SortAToZOutlined';
import { getColumnKey, getColumnPos, renderColumnTitle } from '../utils';
import { TableSort } from '../interface';
import { IconButton } from '../../button';
function getMultiplePriority(column) {
  if (typeof column.sorter === 'object' && typeof column.sorter.multiple === 'number') {
    return column.sorter.multiple;
  }
  return false;
}
function getSortFunction(sorter) {
  if (typeof sorter === 'function') {
    return sorter;
  }
  if (sorter && typeof sorter === 'object' && sorter.compare) {
    return sorter.compare;
  }
  return false;
}
function nextSortDirection(sortDirections, current) {
  if (!current) {
    return sortDirections[0];
  }
  return sortDirections[sortDirections.indexOf(current) + 1];
}
function collectSortStates(columns, init, pos) {
  let sortStates = [];
  function pushState(column, columnPos) {
    var _column$sortOrder;
    sortStates.push({
      column,
      key: getColumnKey(column, columnPos),
      multiplePriority: getMultiplePriority(column),
      sortOrder: (_column$sortOrder = column.sortOrder) !== null && _column$sortOrder !== void 0 ? _column$sortOrder : null
    });
  }
  (columns || []).forEach((column, index) => {
    const columnPos = getColumnPos(index, pos);
    if (column.children) {
      if ('sortOrder' in column) {
        // Controlled
        pushState(column, columnPos);
      }
      sortStates = [...sortStates, ...collectSortStates(column.children, init, columnPos)];
    } else if (column.sorter) {
      if ('sortOrder' in column) {
        // Controlled
        pushState(column, columnPos);
      } else if (init && column.defaultSortOrder) {
        // Default sorter
        sortStates.push({
          column,
          key: getColumnKey(column, columnPos),
          multiplePriority: getMultiplePriority(column),
          sortOrder: column.defaultSortOrder
        });
      }
    }
  });
  return sortStates;
}
function stateToInfo(sorterStates) {
  const {
    column,
    sortOrder
  } = sorterStates;
  return {
    column,
    order: sortOrder,
    field: column.dataIndex,
    columnKey: column.key
  };
}
function generateSorterInfo(sorterStates) {
  const list = sorterStates.filter(_ref => {
    let {
      sortOrder
    } = _ref;
    return sortOrder;
  }).map(stateToInfo);
  if (list.length <= 1) {
    return list[0] || {};
  }
  return list;
}
export function getSortData(data, sortStates, childrenColumnName) {
  const innerSorterStates = sortStates.slice().sort((a, b) => b.multiplePriority - a.multiplePriority);
  const cloneData = data.slice();
  const runningSorters = innerSorterStates.filter(_ref2 => {
    let {
      column: {
        sorter
      },
      sortOrder
    } = _ref2;
    return getSortFunction(sorter) && sortOrder;
  });

  // Skip if no sorter needed
  if (!runningSorters.length) {
    return cloneData;
  }
  return cloneData.sort((record1, record2) => {
    for (let i = 0; i < runningSorters.length; i += 1) {
      const sorterState = runningSorters[i];
      const {
        column: {
          sorter
        },
        sortOrder
      } = sorterState;
      const compareFn = getSortFunction(sorter);
      if (compareFn && sortOrder) {
        const compareResult = compareFn(record1, record2, sortOrder);
        if (compareResult !== 0) {
          return sortOrder === TableSort.ASCEND ? compareResult : -compareResult;
        }
      }
    }
    return 0;
  }).map(record => {
    const subRecords = record[childrenColumnName];
    if (subRecords) {
      return {
        ...record,
        [childrenColumnName]: getSortData(subRecords, sortStates, childrenColumnName)
      };
    }
    return record;
  });
}
export default function useSorter(_ref3) {
  let {
    prefix,
    mergedColumns,
    onSorterChange,
    sortDirections: tableSortDirections,
    tableLocale,
    showSorterTooltip: tableShowSorterTooltip
  } = _ref3;
  const [sortStates, setSortStates] = useState(collectSortStates(mergedColumns, true));
  const mergedSorterStates = useMemo(() => {
    let validate = true;
    const collectedStates = collectSortStates(mergedColumns, false);
    if (!collectedStates.length) {
      return sortStates;
    }
    const validateStates = [];
    function patchStates(state) {
      if (validate) {
        validateStates.push(state);
      } else {
        validateStates.push({
          ...state,
          sortOrder: null
        });
      }
    }
    let multipleMode = null;
    collectedStates.forEach(state => {
      if (multipleMode === null) {
        patchStates(state);
        if (state.sortOrder) {
          if (state.multiplePriority === false) {
            validate = false;
          } else {
            multipleMode = true;
          }
        }
      } else if (multipleMode && state.multiplePriority !== false) {
        patchStates(state);
      } else {
        validate = false;
        patchStates(state);
      }
    });
    return validateStates;
  }, [mergedColumns, sortStates]);

  // Get render columns title required props
  const columnTitleSorterProps = useMemo(() => {
    var _sortColumns$, _sortColumns$2;
    const sortColumns = mergedSorterStates.map(_ref4 => {
      let {
        column,
        sortOrder
      } = _ref4;
      return {
        column,
        order: sortOrder
      };
    });
    return {
      sortColumns,
      sortColumn: (_sortColumns$ = sortColumns[0]) === null || _sortColumns$ === void 0 ? void 0 : _sortColumns$.column,
      sortOrder: (_sortColumns$2 = sortColumns[0]) === null || _sortColumns$2 === void 0 ? void 0 : _sortColumns$2.order
    };
  }, [mergedSorterStates]);
  const triggerSorter = useCallback(sortState => {
    let newSorterStates;
    if (sortState.multiplePriority === false || !mergedSorterStates.length || mergedSorterStates[0].multiplePriority === false) {
      newSorterStates = [sortState];
    } else {
      newSorterStates = [...mergedSorterStates.filter(_ref5 => {
        let {
          key
        } = _ref5;
        return key !== sortState.key;
      }), sortState];
    }
    setSortStates(newSorterStates);
    onSorterChange(generateSorterInfo(newSorterStates), newSorterStates);
  }, [mergedSorterStates, onSorterChange]);
  const transformColumns = useCallback(innerColumns => {
    const injectSorter = (columns, sorterSates, pos) => columns.map((column, index) => {
      const columnPos = getColumnPos(index, pos);
      let newColumn = column;
      if (newColumn.sorter) {
        var _newColumn$showSorter;
        const sortDirections = newColumn.sortDirections || tableSortDirections;
        const showSorterTooltip = (_newColumn$showSorter = newColumn.showSorterTooltip) !== null && _newColumn$showSorter !== void 0 ? _newColumn$showSorter : tableShowSorterTooltip;
        const columnKey = getColumnKey(newColumn, columnPos);
        const sorterState = sorterSates.find(_ref6 => {
          let {
            key
          } = _ref6;
          return key === columnKey;
        });
        const sorterOrder = sorterState ? sorterState.sortOrder : null;
        const nextSortOrder = nextSortDirection(sortDirections, sorterOrder);
        let upNode = null;
        let downNode = null;
        if (sortDirections.includes(TableSort.ASCEND) && sortDirections.includes(TableSort.DESCEND)) {
          upNode = /*#__PURE__*/React.createElement(ExpandUpFilled, {
            className: clsx(`${prefix}-column-sorter-up`, {
              active: sorterOrder === TableSort.ASCEND
            })
          });
          downNode = /*#__PURE__*/React.createElement(ExpandDownFilled, {
            className: clsx(`${prefix}-column-sorter-down`, {
              active: sorterOrder === TableSort.DESCEND
            })
          });
        } else {
          if (sortDirections.includes(TableSort.ASCEND)) {
            upNode = /*#__PURE__*/React.createElement(IconButton, {
              color: sorterOrder === TableSort.ASCEND ? 'primary' : undefined,
              size: "small",
              selected: sorterOrder === TableSort.ASCEND
              // info={sorterOrder === TableSort.ASCEND}
              ,
              icon: /*#__PURE__*/React.createElement(SortAToZOutlined, null)
            });
          }
          if (sortDirections.includes(TableSort.DESCEND)) {
            downNode = /*#__PURE__*/React.createElement(IconButton, {
              color: sorterOrder === TableSort.DESCEND ? 'primary' : undefined,
              size: "small",
              selected: sorterOrder === TableSort.DESCEND
              // info={sorterOrder === TableSort.DESCEND}
              ,
              icon: /*#__PURE__*/React.createElement(SortZToAOutlined, null)
            });
          }
        }
        const {
          cancelSort,
          triggerAsc,
          triggerDesc
        } = tableLocale || {};
        let sortTip = cancelSort;
        if (nextSortOrder === TableSort.DESCEND) {
          sortTip = triggerDesc;
        } else if (nextSortOrder === TableSort.ASCEND) {
          sortTip = triggerAsc;
        }
        newColumn = {
          ...newColumn,
          className: clsx(newColumn.className, {
            [`${prefix}-column-sort`]: sorterOrder
          }),
          title: renderProps => {
            const renderSortTitle = /*#__PURE__*/React.createElement(React.Fragment, null, renderColumnTitle(column.title, renderProps), /*#__PURE__*/React.createElement(Tooltip, {
              title: sortTip,
              disabled: !showSorterTooltip
            }, /*#__PURE__*/React.createElement("span", {
              className: clsx(`${prefix}-column-sorter`, {
                [`${prefix}-column-sorter-full`]: upNode && downNode
              })
            }, /*#__PURE__*/React.createElement("span", {
              className: `${prefix}-column-sorter-inner`
            }, upNode, downNode))));
            return renderSortTitle;
          },
          onHeaderCell: col => {
            const cell = column.onHeaderCell && column.onHeaderCell(col) || {};
            const originOnClick = cell.onClick;
            cell.onClick = event => {
              triggerSorter({
                column,
                key: columnKey,
                sortOrder: nextSortOrder,
                multiplePriority: getMultiplePriority(column)
              });
              if (originOnClick) {
                originOnClick(event);
              }
            };
            cell.className = clsx(cell.className, `${prefix}-column-has-sorters`);
            return cell;
          }
        };
      }
      if ('children' in newColumn) {
        newColumn = {
          ...newColumn,
          children: injectSorter(newColumn.children, sorterSates, columnPos)
        };
      }
      return newColumn;
    });
    return injectSorter(innerColumns, mergedSorterStates);
  }, [mergedSorterStates, tableSortDirections, tableShowSorterTooltip, prefix, tableLocale, triggerSorter]);
  const getSorters = useCallback(() => generateSorterInfo(mergedSorterStates), [mergedSorterStates]);
  return [transformColumns, mergedSorterStates, columnTitleSorterProps, getSorters];
}
