// @universe-design/react/es/transfer/hooks/TransferContext.js
import React from 'react';
import { useCssScope } from '../../config-provider';
import { createContextHook, useControlledState, useConstant } from '@universe-design/react-utilities';
import { useTransferData } from './useTransferData';
import { TransferChangeAction } from '../interface';
import { defaultFilterOption } from '../utils';

/**
 * @public
 */

export function useTransfer(props) {
  const {
    value: valueProp,
    defaultValue,
    options = [],
    onChange,
    onClear,
    onSearch,
    onSelect,
    onDeSelect,
    onToggleSelectAll,
    filterOption = defaultFilterOption,
    group = false,
    defaultSearchValue
  } = props;
  const cssScope = useCssScope('transfer');
  const [searchValue, _setSearchValue] = React.useState(defaultSearchValue || '');
  const [value, _setValue] = useControlledState(defaultValue !== null && defaultValue !== void 0 ? defaultValue : [], valueProp, {
    isControlledOverride: 'value' in props
  });
  const selectedCache = useConstant(() => new Map());
  const {
    optionsMap,
    filteredOptions,
    groupWeakMap
  } = useTransferData({
    options,
    value,
    searchValue,
    filterOption,
    group
  });
  const setValue = React.useCallback((newValue, action) => {
    newValue.forEach(v => {
      const option = optionsMap.get(v);
      if (!option) {
        return;
      }
      // 实时更新 optionsMap 防止下次 render 时 options 被变更
      selectedCache.set(v, option);
    });
    const _options = newValue.map(v => selectedCache.get(v));
    onChange === null || onChange === void 0 ? void 0 : onChange(newValue, {
      action,
      options: _options
    });
    _setValue(newValue);
  }, [_setValue, onChange, optionsMap, selectedCache]);
  const valueSet = React.useMemo(() => new Set(value), [value]);
  React.useMemo(() => {
    // 每次 render 把 value 值对应的 options 存进 selectedCache
    for (const o of valueSet) {
      const option = optionsMap.get(o);
      const cache = option ||
      // 优先使用当前 options 里面的数据
      selectedCache.get(o) ||
      // 其次如果之前缓存过，仍使用之前缓存的数据
      {
        value: o
      }; // 最后，用 value 构造一个 option
      selectedCache.set(o, cache);
    }
  }, [optionsMap, selectedCache, valueSet]);
  /**
   * 已选选项数据
   */
  const selectedOptions = React.useMemo(() => {
    const selected = [];
    valueSet.forEach(v => {
      const cached = selectedCache.get(v);
      if (cached) {
        selected.push(cached);
      } else {
        // 对异常的处理：可能存在 value 不在 option 里面的情况
        selected.push({
          value: v
        });
      }
    });
    return selected;
  }, [selectedCache, valueSet]);
  const noneDisableKeys = React.useMemo(() => filteredOptions.filter(option => !option.disabled).map(o => o.value), [filteredOptions]);
  const setSearchValue = React.useCallback(v => {
    onSearch === null || onSearch === void 0 ? void 0 : onSearch(v);
    _setSearchValue(v);
  }, [onSearch]);

  /**
   * 反选
   */
  const toggle = React.useCallback(v => {
    let newValue;
    /**
     * 可能在删除已选值时，已选值不在 options 里面
     * 这里要从 optionsMap、selectedCache 分别尝试取一下
     **/
    const option = optionsMap.get(v) || selectedCache.get(v);
    if (!option) {
      return;
    }
    if (valueSet.has(v)) {
      newValue = value.filter(k => k !== v);
      onDeSelect === null || onDeSelect === void 0 ? void 0 : onDeSelect(v, {
        option
      });
      setValue(newValue, TransferChangeAction.DeSelect);
    } else {
      onSelect === null || onSelect === void 0 ? void 0 : onSelect(v, {
        option
      });
      newValue = [...value, v];
      setValue(newValue, TransferChangeAction.Select);
    }
  }, [optionsMap, valueSet, value, onDeSelect, setValue, onSelect, selectedCache]);
  const isAllSelected = React.useMemo(() => noneDisableKeys.every(key => valueSet.has(key)) && noneDisableKeys.length > 0, [noneDisableKeys, valueSet]);

  /**
   * 清除所有
   */
  const clearAll = React.useCallback(() => {
    onClear === null || onClear === void 0 ? void 0 : onClear();
    setValue(value.filter(ele => {
      var _optionsMap$get;
      return (_optionsMap$get = optionsMap.get(ele)) === null || _optionsMap$get === void 0 ? void 0 : _optionsMap$get.disabled;
    }), TransferChangeAction.ClearAll);
  }, [onClear, optionsMap, setValue, value]);

  /**
   * 反选所有
   */
  const toggleSelectAll = React.useCallback(() => {
    if (isAllSelected) {
      onToggleSelectAll === null || onToggleSelectAll === void 0 ? void 0 : onToggleSelectAll(false);
      setValue(value.filter(ele => !noneDisableKeys.includes(ele)), TransferChangeAction.DeSelectAll);
    } else {
      onToggleSelectAll === null || onToggleSelectAll === void 0 ? void 0 : onToggleSelectAll(true);
      noneDisableKeys.forEach(v => {
        valueSet.add(v);
      });
      setValue([...valueSet], TransferChangeAction.SelectAll);
    }
  }, [isAllSelected, noneDisableKeys, onToggleSelectAll, setValue, value, valueSet]);
  return {
    value,
    setValue,
    selectedOptions,
    toggleSelectAll,
    cssScope,
    toggle,
    searchValue,
    setSearchValue,
    valueSet,
    isAllSelected,
    clearAll,
    filteredOptions,
    groupWeakMap,
    group
  };
}
const {
  useContext: _useTransferContext,
  Provider: _TransferProvider
} = createContextHook('TransferContext', {
  strict: true
});
export const useTransferContext = _useTransferContext;
export const TransferProvider = _TransferProvider;
