// @universe-design/react/src/shared/components/trigger/modifiers/useTriggerObserver.tsx
import type React from 'react';
import { useState, useCallback } from 'react';
import { unstable_batchedUpdates } from 'react-dom';
import { alignElement } from 'dom-align';
import { isElement } from '../@popperjs/core/lib/dom-utils/instanceOf';
import listScrollParents from '../@popperjs/core/lib/dom-utils/listScrollParents';

import type { UDAlignType, RcTriggerType, RectInfo } from '../types';
import type { SizeInfo } from '../../../../hooks';
import { useIsomorphicLayoutEffect as useLayoutEffect } from '../../../../hooks';
import isEqual from 'lodash/isEqual';

interface ObserverConfig {
  popupAlign: UDAlignType;
  triggerRef: React.RefObject<RcTriggerType | null>;
  readyObserver: boolean;

  enablePopupOverflowAuto?: boolean;
  enableforcePopupAlign?: boolean;
}

const defaultSize: SizeInfo = {
  width: 0,
  height: 0,
  top: 0,
  left: 0,
  right: 0,
  bottom: 0,
  x: 0,
  y: 0,
};

const getPositionSize = (rect: DOMRect) => {
  // Safari 浏览器下 rect 尺寸小数点后几位会不稳定
  const left = Math.floor(rect.left);
  const top = Math.floor(rect.top);
  const width = Math.floor(rect.width);
  const height = Math.floor(rect.height);
  return {
    left,
    top,
    x: left,
    y: top,
    width,
    height,
    right: width + left,
    bottom: height + top,
  };
};

/**
 * dom-align 用于获取可视区域的函数
 */
const getVisibleRectForElement: (
  element: Element,
  alwaysByViewport: boolean,
) => { top: number; left: number; bottom: number; right: number } | null = (alignElement as any)
  .__getVisibleRectForElement;

export const useTriggerObserver = (config: ObserverConfig) => {
  const {
    popupAlign,
    triggerRef,
    readyObserver,
    enablePopupOverflowAuto = false,
    enableforcePopupAlign = false,
  } = config;

  const [popupSize, setPopupSize] = useState<SizeInfo>(defaultSize);
  const [childSize, setChildSize] = useState<SizeInfo>(defaultSize);

  const [viewPointRect, setViewPointRect] = useState<RectInfo | null>(null);

  const updateRect = useCallback(() => {
    const child = (triggerRef.current as any)?.getRootDomNode();
    const popup: HTMLElement = (triggerRef.current as any)?.getPopupDomNode();

    if (!child || !popup) {
      return;
    }

    unstable_batchedUpdates(() => {
      const newPopupSize = getPositionSize(popup.getBoundingClientRect());
      const newChildSize = getPositionSize(child.getBoundingClientRect());

      if (!isEqual(newPopupSize, popupSize)) {
        setPopupSize(newPopupSize);
      }

      if (!isEqual(newChildSize, childSize)) {
        setChildSize(newChildSize);
      }

      if (enablePopupOverflowAuto) {
        // FIXME 现在这里逻辑还有点问题，如果 visible rect 大小变了其实我们是无法得到监听的
        const alwaysByViewport = Boolean(popupAlign.overflow?.alwaysByViewport);
        const newViewPointRect = getVisibleRectForElement(popup, alwaysByViewport);

        if (!isEqual(newViewPointRect, viewPointRect)) {
          setViewPointRect(newViewPointRect);
        }
      }
    });
  }, [popupAlign, triggerRef, popupSize, childSize, viewPointRect, enablePopupOverflowAuto]);

  useLayoutEffect(() => {
    if (!readyObserver) {
      return;
    }

    const passive = { passive: true };

    const child: HTMLElement = (triggerRef.current as any)?.getRootDomNode();
    const popup: HTMLElement = (triggerRef.current as any)?.getPopupDomNode();

    if (!child || !popup) {
      return;
    }

    updateRect();

    const forceAlign = () => {
      (triggerRef.current as any)?.forcePopupAlign();
    };

    let scrollParents: HTMLElement[] = [];

    if (enableforcePopupAlign) {
      scrollParents = [
        ...new Set([
          ...(isElement(child) ? listScrollParents(child as Element) : []),
          ...(isElement(popup) ? listScrollParents(popup as Element) : []),
        ]),
      ] as HTMLElement[];

      scrollParents.forEach(scrollParent => {
        scrollParent.addEventListener('scroll', forceAlign, passive);
      });
      window.addEventListener('resize', forceAlign, passive);
    }

    // 观察器的配置（需要观察什么变动）
    const mutationConfig = {
      attributes: true,
      childList: false,
      subtree: false,
      attributeOldValue: true,
      attributeFilter: ['style'],
    };

    // 创建一个观察器实例并传入回调函数
    const mutationObserver = new MutationObserver(mutationsList => {
      // 如果定位发生变化，重新计算位置
      const observeStyleKey = ['left', 'top', 'bottm', 'right'];

      const oldValue = new Map();

      // Use traditional 'for loops' for IE 11
      for (const mutation of mutationsList) {
        if (mutation.type === 'attributes' && mutation.attributeName === 'style') {
          const trim = (str: string) => str.trim();

          if (typeof mutation.oldValue === 'string') {
            const values = mutation.oldValue
              .split(';')
              .map(trim)
              .filter(Boolean)
              .map(e => e.split(':').map(trim));

            values.forEach(([key, value]) => {
              if (!observeStyleKey.includes(key)) {
                return;
              }

              if (oldValue.has(key)) {
                return;
              }

              oldValue.set(key, value);
            });
          }
        }
      }

      let isChanged = false;

      oldValue.forEach((value, key) => {
        if (value !== popup.style[key]) {
          isChanged = true;
        }
      });

      if (isChanged) {
        updateRect();
      }
    });

    mutationObserver.observe(popup, mutationConfig);

    return () => {
      if (enableforcePopupAlign) {
        scrollParents.forEach(scrollParent => {
          scrollParent.removeEventListener('scroll', forceAlign, !passive.passive);
        });
        window.removeEventListener('resize', forceAlign, !passive.passive);
      }

      mutationObserver.disconnect();
    };
  }, [readyObserver, updateRect, triggerRef, enableforcePopupAlign]);

  return [childSize, popupSize, viewPointRect] as const;
};
