// @universe-design/react/src/shared/components/trigger/modifiers/useTriggerPopupOverflowAuto.tsx
import type { CSSProperties } from 'react';
import type React from 'react';
import { cloneElement, isValidElement, useCallback, useMemo, useState } from 'react';
import type { RcTriggerPropsWithRef, UDAlignType } from '../types';
import { alignElement } from 'dom-align';
import type { SizeInfo } from '../../../../hooks';
import { useIsomorphicLayoutEffect as useLayoutEffect } from '../../../../hooks';
import { isFunction, isNonNullable } from '../../../../util';
import getWindow from '../@popperjs/core/lib/dom-utils/getWindow';

/**
 * dom-align 用于获取可视区域的函数
 */
const getVisibleRectForElement: (
  element: Element,
  alwaysByViewport: boolean,
) => { top: number; left: number; bottom: number; right: number } | null = (alignElement as any)
  .__getVisibleRectForElement;

const passive = { passive: true };

/**
 * 已监听：
 * window scroll, resize
 * scrollParents scroll
 * popup resize
 * trigger resize
 * 未监听：
 * scrollParents resize
 */
export const useTriggerPopupOverflowAuto = (
  props: RcTriggerPropsWithRef,
  {
    popup,
    popupSize,
    popupAlign,
    childSize,
    scrollParents,
    scroll,
    resize,
    disabled,
  }: {
    popup: HTMLElement | null;
    popupSize: SizeInfo;
    popupAlign: UDAlignType;
    childSize: SizeInfo;
    scrollParents: (Element | typeof window | VisualViewport)[];
    scroll: boolean;
    resize: boolean;
    disabled: boolean;
  },
): [RcTriggerPropsWithRef] => {
  const [rect, setRect] = useState<{
    top: number;
    left: number;
    right: number;
    bottom: number;
  } | null>(null);

  const updateRect = useCallback(() => {
    // FIXME 现在这里逻辑还有点问题，如果 visible rect 大小变了其实我们是无法得到监听的
    const alwaysByViewport = Boolean(popupAlign.overflow?.alwaysByViewport);
    setRect(popup && getVisibleRectForElement(popup, alwaysByViewport));
  }, [popup, popupAlign.overflow?.alwaysByViewport]);

  useLayoutEffect(() => {
    updateRect();
  }, [
    updateRect,
    childSize.bottom,
    childSize.left,
    childSize.right,
    childSize.top,
    popupSize.bottom,
    popupSize.left,
    popupSize.right,
    popupSize.top,
  ]);

  const popupContentStyle = useMemo<CSSProperties | undefined>(() => {
    const style: CSSProperties = {};
    if (!popup || disabled) {
      return undefined;
    }
    // 如果两个方向都要翻转就不做 maxWidth 和 maxHeight 的计算了
    if (popupAlign.overflow?.adjustX && popupAlign.overflow?.adjustY) {
      return undefined;
    }
    if (!rect) {
      return undefined;
    }
    style.overflow = 'auto';
    if (!popupAlign.overflow?.adjustX) {
      if (childSize.right < popupSize.left) {
        style.maxWidth = rect.right - popupSize.left;
      } else if (childSize.left > popupSize.right) {
        style.maxWidth = popupSize.right - rect.left;
      }
    }
    if (!popupAlign.overflow?.adjustY) {
      if (childSize.bottom < popupSize.top) {
        style.maxHeight = rect.bottom - popupSize.top;
      } else if (childSize.top > popupSize.bottom) {
        style.maxHeight = popupSize.bottom - rect.top;
      }
    }
    return style;
  }, [
    childSize.bottom,
    childSize.left,
    childSize.right,
    childSize.top,
    disabled,
    popup,
    popupAlign.overflow?.adjustX,
    popupAlign.overflow?.adjustY,
    popupSize.bottom,
    popupSize.left,
    popupSize.right,
    popupSize.top,
    rect,
  ]);

  const window = getWindow(popup);

  useLayoutEffect(() => {
    if (disabled) {
      return;
    }
    if (scroll) {
      scrollParents.forEach(scrollParent => {
        scrollParent.addEventListener('scroll', updateRect, passive);
      });
    }

    if (resize) {
      window.addEventListener('resize', updateRect, passive);
    }
    return () => {
      if (scroll) {
        scrollParents.forEach(scrollParent => {
          scrollParent.removeEventListener('scroll', updateRect, !passive.passive);
        });
      }

      if (resize) {
        window.removeEventListener('resize', updateRect, !passive.passive);
      }
    };
  }, [disabled, resize, scroll, scrollParents, updateRect, window]);

  const popupProps = isFunction(props.popup) ? props.popup() : props.popup;

  return [
    {
      ...props,
      popup:
        isValidElement<{ style?: React.CSSProperties }>(popupProps) && isNonNullable(popupContentStyle)
          ? cloneElement(popupProps, {
              style: { ...popupContentStyle, ...popupProps.props.style },
            })
          : popupProps,
    },
  ];
};
