// @universe-design/react/src/button/stories/homeless.stories.tsx
import type { ReactNode } from 'react';
import React, { useState } from 'react';
import type { ButtonProps } from '../..';
import {
  Button,
  IconButton,
  pickIconButtonProps,
  Space,
  Switch,
  Table,
  toast,
  ColorPicker,
  Dropdown,
  Upload,
} from '../..';
import CloseOutlined from '@universe-design/icons-react/CloseOutlined';
import AddOutlined from '@universe-design/icons-react/AddOutlined';
import EditOutlined from '@universe-design/icons-react/EditOutlined';
import SearchOutlineOutlined from '@universe-design/icons-react/SearchOutlineOutlined';
import DownloadOutlined from '@universe-design/icons-react/DownloadOutlined';
import MoreBoldOutlined from '@universe-design/icons-react/MoreBoldOutlined';
import './index.less';

export default {
  title: 'Components/Button/临时示例', // 注意：这里需要修改成自己对应 Component name。
  component: Button,
};

export { default as Basic } from '../demo/basic';
export { default as Text } from '../demo/text';
export { default as Icon } from '../demo/icon';
export { default as Loading } from '../demo/loading';
export { default as Block } from '../demo/block';
export { default as Disabled } from '../demo/disabled';
export { default as Link } from '../demo/link';
export { default as PrimaryScreenshot } from '../demo/z.primary.screenshot';
export { default as PrimaryDangerScreenshot } from '../demo/z.primary-danger.screenshot';
export { default as SecondaryScreenshot } from '../demo/z.secondary.screenshot';
export { default as SecondaryInfoScreenshot } from '../demo/z.secondary-info.screenshot';
export { default as SecondaryDangerScreenshot } from '../demo/z.secondary-danger.screenshot';
export { default as IconSecondaryScreenshot } from '../demo/z.icon-secondary.screenshot';
export { default as TextScreenshot } from '../demo/z.text.screenshot';
export { default as IconScreenshot } from '../demo/z.icon.screenshot';
export { default as RoundScreenshot } from '../demo/z.round.screenshot';

const Sizes = {
  'extra-large': '超大 高48px, 字体18px',
  large: '大 高40px, 字体16px',
  medium: '中 高32px, 字体14px',
  small: '小 高28px, 字体12px',
  'extra-small': '超小 高24px 字体12px',
};

type ExtraProps = Omit<ButtonProps, 'ref'>;

// const shapes = ['', 'round'];

function renderRowRaius(props: ExtraProps, renderer: (props: ExtraProps) => React.ReactNode) {
  return (
    <>
      {['round', ''].map((shape: any, i) => (
        <React.Fragment key={shape}>{renderer({ ...props, shape })}</React.Fragment>
      ))}
    </>
  );
}

const onClick = (e: any) => {
  toast.info(`${e.target.textContent} clicked!`);
};

function renderNormalButton(record: { size: ButtonProps['size'] }, props: ExtraProps): ReactNode {
  function renderRowType(iProps?: ExtraProps, desc = '') {
    return (
      <>
        <Space>
          <Button {...iProps} {...props} onClick={onClick}>
            {desc}
          </Button>
          <Button {...iProps} {...props}>
            取消
          </Button>
          <Button {...iProps} {...props} icon={<EditOutlined />}>
            编辑
          </Button>
          <Button {...iProps} {...props} color="primary" icon={<EditOutlined />}>
            编辑
          </Button>
          <Button {...iProps} {...props} color="danger" icon={<AddOutlined />}>
            新增
          </Button>
          <Button {...iProps} {...props} shape="square" icon={<DownloadOutlined />} />
          <Button {...iProps} {...props} color="primary" shape="square" icon={<EditOutlined />} />
          <Button {...iProps} {...props} color="danger" shape="circle" icon={<AddOutlined />} />
          <Button {...iProps} {...props} color="primary" icon={<DownloadOutlined />}>
            下载
          </Button>
          {/* <Button {...iProps} {...props} color="primary" type="filled">
            确定
          </Button>
          <Button {...iProps} {...props} color="danger" type="filled">
            删除
          </Button>
          <Button
            {...iProps}
            {...props}
            color="primary"
            type="filled"
            shape="square"
            icon={<EditOutlined />}
          />
          <Button
            {...iProps}
            {...props}
            color="danger"
            type="filled"
            shape="circle"
            icon={<AddOutlined />}
          /> */}
        </Space>
      </>
    );
  }

  function renderRow(iProps?: ExtraProps) {
    return [
      renderRowType({ type: 'filled', ...iProps }, 'filled(primary)类型'),
      renderRowType({ type: 'outlined', ...iProps }, 'outlined(default)类型'),
    ];
  }
  return renderRowRaius({ ...record }, renderRow);
}

function renderTextButton(record: { size: ButtonProps['size'] }, props: ExtraProps): ReactNode {
  const commonProps: Omit<ButtonProps, 'ref'> = {
    ...props,
  };

  function renderRow(iProps: ExtraProps) {
    return (
      <Space size="medium">
        <Button {...commonProps} {...iProps} onClick={onClick} icon={<EditOutlined />}>
          更多操作
        </Button>
        <Button {...commonProps} {...iProps}>
          取消
        </Button>
        <Button {...commonProps} {...iProps}>
          确定
        </Button>
        <Button {...commonProps} {...iProps} color="primary">
          确定
        </Button>
        <Button {...commonProps} {...iProps} color="danger" icon={<CloseOutlined />}>
          删除
        </Button>
        <Button {...commonProps} {...iProps} icon={<EditOutlined />} />
        <Button {...commonProps} {...iProps} icon={<EditOutlined />} />
        <Button {...commonProps} {...iProps} icon={<AddOutlined />} />
        <Button {...commonProps} {...iProps} color="primary" icon={<SearchOutlineOutlined />} />
        <Button {...commonProps} {...iProps} color="danger" icon={<CloseOutlined />} />
        <Button {...commonProps} {...iProps} icon={<EditOutlined />} />
        <Button {...commonProps} {...iProps} icon={<AddOutlined />} />
        <Button {...commonProps} {...iProps} color="primary" icon={<SearchOutlineOutlined />} />
        <Button {...commonProps} {...iProps} color="danger" icon={<CloseOutlined />} />
        <Button {...commonProps} {...iProps} color="primary" icon={<SearchOutlineOutlined />} />
        <Button {...commonProps} {...iProps} color="danger" icon={<CloseOutlined />} />
      </Space>
    );
  }

  return renderRowRaius({ ...record }, renderRow);
}

function renderIconButton(record: { size: ButtonProps['size'] }, props: ExtraProps): ReactNode {
  const commonProps: Omit<ButtonProps, 'ref'> = {
    ...props,
    type: 'icon',
  };
  function renderRow(iProps: ExtraProps) {
    return (
      <Space>
        <Button {...commonProps} {...iProps} onClick={onClick} icon={<MoreBoldOutlined />} />
        <Button {...commonProps} {...iProps} icon={<MoreBoldOutlined />} />
        <Button {...commonProps} {...iProps} icon={<EditOutlined />} />
        <Button {...commonProps} {...iProps} icon={<AddOutlined />} />
        <Button {...commonProps} {...iProps} color="primary" icon={<SearchOutlineOutlined />} />
        <Button {...commonProps} {...iProps} color="danger" icon={<CloseOutlined />} />
        <Button {...commonProps} {...iProps} color="primary" icon={<AddOutlined />} />
        <Button {...commonProps} {...iProps} color="danger" icon={<CloseOutlined />} />
        <IconButton
          {...pickIconButtonProps(commonProps)}
          {...pickIconButtonProps(iProps)}
          color="primary"
          icon={<SearchOutlineOutlined />}
        />
        <IconButton
          {...pickIconButtonProps(commonProps)}
          {...pickIconButtonProps(iProps)}
          size={record.size}
          color="danger"
          icon={<CloseOutlined />}
        />
        <IconButton
          {...pickIconButtonProps(commonProps)}
          {...pickIconButtonProps(iProps)}
          size={record.size}
          color="primary"
          icon={<AddOutlined />}
        />
        <IconButton
          {...pickIconButtonProps(commonProps)}
          {...pickIconButtonProps(iProps)}
          size={record.size}
          icon={<CloseOutlined />}
        />
        <IconButton
          {...pickIconButtonProps(commonProps)}
          {...pickIconButtonProps(iProps)}
          size={60}
          color="primary"
          icon={<CloseOutlined />}
        />
        <IconButton
          {...pickIconButtonProps(commonProps)}
          {...pickIconButtonProps(iProps)}
          size={60}
          color="danger"
          icon={<CloseOutlined />}
        />
        <IconButton
          {...pickIconButtonProps(commonProps)}
          {...pickIconButtonProps(iProps)}
          size={60}
          icon={<CloseOutlined />}
        />
      </Space>
    );
  }

  return renderRowRaius({ ...record }, renderRow);
}

function renderBase(record: { size: ButtonProps['size'] }, props: ExtraProps): ReactNode {
  function renderRow(iProps?: ExtraProps) {
    return (
      <Space>
        <Button {...iProps} {...props}>
          更多操作
        </Button>
        <Button {...iProps} {...props}>
          取消
        </Button>
        <Button {...iProps} {...props} icon={<EditOutlined />}>
          编辑
        </Button>
        <Button {...iProps} {...props} color="primary" icon={<EditOutlined />}>
          编辑
        </Button>
        <Button {...iProps} {...props} color="danger" icon={<AddOutlined />}>
          新增
        </Button>
        <Button {...iProps} {...props} shape="square" icon={<DownloadOutlined />} />
        <Button {...iProps} {...props} color="primary" shape="square" icon={<EditOutlined />} />
        <Button {...iProps} {...props} color="danger" shape="circle" icon={<AddOutlined />} />
        <Button {...iProps} {...props} color="primary" icon={<DownloadOutlined />}>
          下载
        </Button>
      </Space>
    );
  }
  return renderRowRaius({ ...record }, renderRow);
}

function renderSpecial(record: { size: ButtonProps['size'] }, props: ExtraProps): ReactNode {
  function renderRow(iProps?: ExtraProps) {
    return (
      <Space>
        <Button {...iProps} {...props}>
          更
        </Button>
        <Button {...iProps} {...props}>
          多
        </Button>
        <Button {...iProps} {...props} icon={<EditOutlined />}>
          编辑长
        </Button>
        <Button {...iProps} {...props} color="primary" icon={<EditOutlined />}>
          编辑
        </Button>
        <Button {...iProps} {...props} color="danger" icon={<AddOutlined />}>
          新增
        </Button>
        <Button {...iProps} {...props} shape="square" icon={<DownloadOutlined />} />
        <Button {...iProps} {...props} color="primary" shape="square" icon={<EditOutlined />} />
        <Button {...iProps} {...props} color="danger" shape="circle" icon={<AddOutlined />} />
        <Button {...iProps} {...props} color="primary" icon={<DownloadOutlined />}>
          下载
        </Button>
        <Button {...iProps} {...props} href="https://www.feishu.com" color="primary" icon={<DownloadOutlined />}>
          LinkButton Download
        </Button>
      </Space>
    );
  }
  function renderRows(iProps?: ExtraProps) {
    const types = ['primary', 'default'];
    const noMinWidth = [true, false];
    const endIcons = [null, <AddOutlined key="icon" />];
    const conf: ExtraProps[] = Array(8)
      .fill(1)
      .map((_, i) => ({
        type: types[(i >> 2) & 0b1] as ExtraProps['type'],
        noMinWidth: noMinWidth[(i >> 1) & 0b1] as ExtraProps['noMinWidth'],
        endIcon: endIcons[(i >> 0) & 0b1] as ExtraProps['endIcon'],
      }));
    return conf.map((c, i) => <React.Fragment key={i}>{renderRow({ ...iProps, ...c })}</React.Fragment>);
  }
  return renderRowRaius({ ...record }, renderRows);
}

export const BaseButtons = (): ReactNode => {
  const [loading, setLoading] = useState(false);
  const [disabled, setDisabled] = useState(false);
  const [, set] = useState(false);
  const [Bg, setBg] = useState<string | undefined>('none');
  const columnOne = {
    title: '样式',
    dataIndex: 'size',
    width: '150px',
    render(txt: keyof typeof Sizes) {
      return (
        <p>
          {txt}
          <br />
          {Sizes[txt]}
        </p>
      );
    },
  };

  const getColumnTwo = (title: string, props: ExtraProps, renderer: typeof renderNormalButton) => ({
    title,
    dataIndex: 'normal',
    render(_: string, record: { size: ButtonProps['size'] }) {
      return renderer(record, { loading, disabled, ...props });
    },
  });

  const columnsNormal = [columnOne, getColumnTwo('普通按钮', {}, renderNormalButton)];
  const columnsIcon = [columnOne, getColumnTwo('图标按钮', { type: 'icon' }, renderIconButton)];
  const columnsText = [columnOne, getColumnTwo('文本按钮', { type: 'text' }, renderTextButton)];

  const columnsTextSecondary = [columnOne, getColumnTwo('link 按钮', { type: 'link' }, renderBase)];

  // const columnsGhost = [
  //   columnOne,
  //   getColumnTwo('幽灵按钮', { type: 'ghost' }, renderBase),
  // ];

  const columnsButtonLink = [
    columnOne,
    getColumnTwo('ButtonLink', { href: 'https://universe.bytedance.net' }, renderBase),
  ];

  const columnsSpecial = [columnOne, getColumnTwo('ButtonLink', {}, renderSpecial)];
  const data: any = [
    { size: 'extra-large' },
    { size: 'large' },
    { size: 'medium' },
    { size: 'small' },
    { size: 'extra-small' },
  ];
  return (
    <div className="ud-btn-demo" style={{ background: Bg }}>
      <section>
        <Space>
          <span>
            Loading：
            <Switch
              onChange={() => {
                setLoading(!loading);
              }}
            />
          </span>
          <span>
            Disabled：
            <Switch
              onChange={() => {
                setDisabled(!disabled);
              }}
            />
          </span>
          <span>
            ：
            <Switch
              onChange={() => {
                set(false);
              }}
            />
          </span>
          <span>
            页面背景色：
            <Dropdown
              overlay={
                <ColorPicker
                  onChange={color => {
                    setBg(color);
                  }}
                />
              }
            >
              <Button>选择颜色</Button>
            </Dropdown>
          </span>
        </Space>
      </section>
      <section>
        <h2>普通按钮</h2>
        <Table rowKey={o => o.size as string} columns={columnsNormal} dataSource={data} />
      </section>
      <section>
        <h2>普通 Icon 按钮</h2>
        <p>
          这类型的 Icon 按钮使用了给定的 5 个尺寸，和文本按钮中的纯 Icon
          按钮不同，如果这里尺寸没有符合需求的按钮，可以使用 Text 按钮。
        </p>
        <Table rowKey={o => o.size as string} columns={columnsIcon} dataSource={data} />
      </section>
      <section>
        <h2>Text 按钮</h2>
        <p>
          文本按钮的按钮尺寸是根据字体大小来设定的，包含文本的按钮 hover 时，背景大小为行高加固定的 4px padding，
          <br />
          纯 Icon 按钮 hover 时，背景大小为 Icon 大小加固定的 4px padding。
          <br />
          你可以通过覆盖 font-size 来实现更多的尺寸需求。 font-size 为 10px 时，在 chrome 上默认时会使用 12px， 但是 svg
          icon 时可以缩小到 10px 的。
        </p>
        <Table rowKey={o => o.size as string} columns={columnsText} dataSource={data} />
      </section>
      {/* <section>
        <h2>幽灵按钮</h2>
        <p>主要应用于深色背景或图片背景上，通过点击按钮触发即时操作。</p>
        <Table
          rowKey={o => o.size as string}
          columns={columnsGhost}
          dataSource={data}
          style={{ background: '#000', color: '#fff' }}
        />
      </section> */}
      <section>
        <h2>Link</h2>
        <Space>
          <Button type="link" href="#">
            文字链
          </Button>
          <Button type="link" href="#">
            文字链
          </Button>
          <Button type="link" loading href="#">
            文字链
          </Button>
          <Button type="link" disabled href="#">
            文字链
          </Button>
        </Space>
      </section>
      <section>
        <h2>Button Link</h2>
        <p>具有 Link 外观的按钮</p>
        <Table rowKey={o => o.size as string} columns={columnsTextSecondary} dataSource={data} />
      </section>
      <section>
        <h2>Link 按钮(Link Button)</h2>
        <p>具有 Button 外观的链接，通过设置 href 属性触发</p>
        <Table rowKey={o => o.size as string} columns={columnsButtonLink} dataSource={data} />
      </section>
      <section>
        <h2>特殊 Case</h2>
        <Table rowKey={o => o.size as string} columns={columnsSpecial} dataSource={data} />
      </section>
    </div>
  );
};

export const WithUpload = () => (
  <p>
    <p>点击 disabled 态 iconButton 不应该触发 upload 的选择框</p>
    <p>Button 应当阻止内部的 click 冒泡</p>
    <Upload action="xxx" disabled>
      <IconButton icon={<EditOutlined />} disabled />
    </Upload>
  </p>
);
