// @universe-design/react/src/button/stories/NormalButtonDemo.tsx
/* eslint-disable no-nested-ternary */
/* eslint-disable max-nested-callbacks */
import React from 'react';
import { Button, Space } from '@universe-design/react';
import AddOutlined from '@universe-design/icons-react/AddOutlined';
import type { ButtonProps } from '../ButtonBase';

export const ButtonDemoContainer: React.FC<React.PropsWithChildren<{ title?: string }>> = ({ title, children }) => (
  <Space style={{ display: 'flex', marginBottom: 24 }}>
    <div style={{ width: 120, fontWeight: 'bold', textAlign: 'center' }}>{title}</div>
    <Space
      direction="vertical"
      style={{
        flex: 1,
        border: '1px dashed #7B61FF',
        padding: 20,
        borderRadius: 5,
      }}
      size={16}
    >
      {children}
    </Space>
  </Space>
);

export const ButtonDemoRow: React.FC<React.PropsWithChildren<unknown>> = ({ children }) => (
  <Space size={0}>{children}</Space>
);

export const NormalButtonDemo = ({
  fixedWidthDemo = true,
  buttonName = '主要按钮',
  cases = [{}, { loading: true }, { disabled: true }],
  ...props
}: ButtonProps & {
  fixedWidthDemo?: boolean;
  buttonName?: string;
  cases?: ButtonProps[];
}) => (
  <Space style={{ display: 'flex' }} direction="vertical">
    {
      // eslint-disable-next-line react/jsx-key
      ([null, <AddOutlined />] as const).map(icon => (
        <div key={String(icon)}>
          {(
            [
              { size: 'extra-small', content: '超小按钮' },
              { size: 'small', content: '小号按钮' },
              { size: 'medium', content: buttonName },
              { size: 'large', content: '大号按钮' },
              { size: 'extra-large', content: '超大按钮' },
            ] as const
          ).map(sizeProps => (
            <ButtonDemoContainer key={sizeProps.size} title={sizeProps.size}>
              {(['adaptive', fixedWidthDemo ? 'fixed' : null] as const).filter(Boolean).map(widthMode => (
                <ButtonDemoRow key={widthMode}>
                  {cases.map(caseProps => (
                    <div key={JSON.stringify(caseProps)} style={{ width: 136 }}>
                      <Button
                        type="primary"
                        icon={icon}
                        size={sizeProps.size}
                        // eslint-disable-next-line react/no-children-prop
                        children={
                          widthMode === 'adaptive' ? sizeProps.content : icon || caseProps.loading ? '好' : '按钮'
                        }
                        {...caseProps}
                        {...props}
                      />
                    </div>
                  ))}
                </ButtonDemoRow>
              ))}
            </ButtonDemoContainer>
          ))}
        </div>
      ))
    }
  </Space>
);
