// @universe-design/react/src/button/stories/screenshot.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import CBasic from '../demo/basic';
import CBlock from '../demo/block';
import CDisabled from '../demo/disabled';
import CIcon from '../demo/icon';
import CLink from '../demo/link';
import CLoading from '../demo/loading';
import CText from '../demo/text';
import CZIconSecondaryScreenshot from '../demo/z.icon-secondary.screenshot';
import CZIconScreenshot from '../demo/z.icon.screenshot';
import CZPrimaryDangerScreenshot from '../demo/z.primary-danger.screenshot';
import CZPrimaryScreenshot from '../demo/z.primary.screenshot';
import CZRoundScreenshot from '../demo/z.round.screenshot';
import CZSecondaryDangerScreenshot from '../demo/z.secondary-danger.screenshot';
import CZSecondaryInfoScreenshot from '../demo/z.secondary-info.screenshot';
import CZSecondaryScreenshot from '../demo/z.secondary.screenshot';
import CZTextScreenshot from '../demo/z.text.screenshot';
if (!(CBasic as Story).storyName) {
  (CBasic as Story).storyName = 'Basic';
}
if (!(CBlock as Story).storyName) {
  (CBlock as Story).storyName = 'Block';
}
if (!(CDisabled as Story).storyName) {
  (CDisabled as Story).storyName = 'Disabled';
}
if (!(CIcon as Story).storyName) {
  (CIcon as Story).storyName = 'Icon';
}
if (!(CLink as Story).storyName) {
  (CLink as Story).storyName = 'Link';
}
if (!(CLoading as Story).storyName) {
  (CLoading as Story).storyName = 'Loading';
}
if (!(CText as Story).storyName) {
  (CText as Story).storyName = 'Text';
}
if (!(CZIconSecondaryScreenshot as Story).storyName) {
  (CZIconSecondaryScreenshot as Story).storyName = 'Z icon secondary screenshot';
}
if (!(CZIconScreenshot as Story).storyName) {
  (CZIconScreenshot as Story).storyName = 'Z icon screenshot';
}
if (!(CZPrimaryDangerScreenshot as Story).storyName) {
  (CZPrimaryDangerScreenshot as Story).storyName = 'Z primary danger screenshot';
}
if (!(CZPrimaryScreenshot as Story).storyName) {
  (CZPrimaryScreenshot as Story).storyName = 'Z primary screenshot';
}
if (!(CZRoundScreenshot as Story).storyName) {
  (CZRoundScreenshot as Story).storyName = 'Z round screenshot';
}
if (!(CZSecondaryDangerScreenshot as Story).storyName) {
  (CZSecondaryDangerScreenshot as Story).storyName = 'Z secondary danger screenshot';
}
if (!(CZSecondaryInfoScreenshot as Story).storyName) {
  (CZSecondaryInfoScreenshot as Story).storyName = 'Z secondary info screenshot';
}
if (!(CZSecondaryScreenshot as Story).storyName) {
  (CZSecondaryScreenshot as Story).storyName = 'Z secondary screenshot';
}
if (!(CZTextScreenshot as Story).storyName) {
  (CZTextScreenshot as Story).storyName = 'Z text screenshot';
}

export default {
  title: 'Components/Button/截图测试',
  tags: ['autodocs'],
};

export {
  CBasic,
  CBlock,
  CDisabled,
  CIcon,
  CLink,
  CLoading,
  CText,
  CZIconSecondaryScreenshot,
  CZIconScreenshot,
  CZPrimaryDangerScreenshot,
  CZPrimaryScreenshot,
  CZRoundScreenshot,
  CZSecondaryDangerScreenshot,
  CZSecondaryInfoScreenshot,
  CZSecondaryScreenshot,
  CZTextScreenshot,
};
