// @universe-design/react-utilities/lib-esm/util/size.js
export const sizeEnum = ['xl', 'lg', 'md', 'sm', 'xs'];
/**
 * @public
 */
export const defaultSizeEnum = ['large', 'medium', 'small'];

/**
 * 预设的一些尺寸
 * @public
 */

export const defaultExtraSizeEnum = ['extra-large', 'extra-small'
// 'default',
];
/**
 * @public
 */

export const sizeMap = {
  'extra-large': 'xl',
  large: 'lg',
  medium: 'md',
  // default: 'md',
  small: 'sm',
  'extra-small': 'xs'
};
export function isValidDefaultSize(sizeName) {
  return defaultSizeEnum.includes(sizeName);
}
export function isValidSize(sizeName) {
  return defaultSizeEnum.includes(sizeName) || defaultExtraSizeEnum.includes(sizeName);
}
export function getDefaultSizeAbbr(sizeName, defaultName) {
  if (isValidDefaultSize(sizeName)) {
    return sizeMap[sizeName];
  }
  if (defaultName && isValidDefaultSize(sizeName)) {
    return sizeMap[defaultName];
  }
  return '';
}
/**
 * @public
 */
export function getSizeAbbr(sizeName) {
  if (isValidSize(sizeName)) {
    return sizeMap[sizeName];
  }
  return undefined;
}
