// @universe-design/react/lib/date-picker/panels/year-body/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _Body = _interopRequireDefault(require("../base/Body"));
var _config = require("../base/config");
var _hooks = require("../../../config-provider/hooks");
var _utils = require("../base/utils");
var _cell = require("../../cell");
var _utils2 = require("../DateCalendar/utils");
var _DateCalendarUIContext = require("../DateCalendar/DateCalendarUIContext");
function YearBody() {
  const prefixCls = (0, _hooks.useCssScope)('picker')();
  const yearPanelPrefixCls = `${prefixCls}-year-panel`;
  const dateCellInteractiveAreaCls = (0, _cell.useDateCellInteractiveAreaCls)();
  const {
    shouldCellDisabled,
    pickerValue,
    cellRenderer,
    previewDate,
    selectedDates,
    selectedRange
  } = (0, _DateCalendarUIContext.useDateCalendarUIContext)();
  return /*#__PURE__*/_react.default.createElement(_Body.default, (0, _extends2.default)({}, _config.BodyConfig.year, {
    className: yearPanelPrefixCls,
    cellClassName: `${yearPanelPrefixCls}-cell`,
    colNum: 4,
    rowNum: 5,
    baseDate: (0, _utils.getYearStart)(pickerValue.clone()),
    getCellStatus: date => {
      const isRangePreview = previewDate ? previewDate(date) : false;
      const [startDate = null, endDate = null] = selectedRange || [];
      return {
        isDisabled: shouldCellDisabled(date, 'year'),
        isRange: date.isSameOrAfter(startDate, 'year') && date.isSameOrBefore(endDate, 'year'),
        isSelected: (0, _utils2.isSelectCell)({
          mode: 'year',
          selectedDates,
          selectedRange,
          date
        }),
        isRangePreview
      };
    },
    getCellDate: (d, offset) => d.clone().add(offset, 'year'),
    dateRender: d => {
      const node = /*#__PURE__*/_react.default.createElement("div", {
        className: dateCellInteractiveAreaCls
      }, d.format('YYYY'));
      if (typeof cellRenderer === 'function') {
        return cellRenderer(d, {
          viewMode: 'year',
          defaultNode: node
        });
      }
      return node;
    }
  }));
}
var _default = exports.default = YearBody;
