// @universe-design/react/es/context-menu/index.js
import React from 'react';
import ReactDOM from 'react-dom';
import { useComplementEvent } from '../hooks/useComplementEvent';
import { useCssScope } from '../config-provider/hooks';
import { isFunction } from '../util';
import { useIsomorphicLayoutEffect as useLayoutEffect } from '../hooks';
import { CrossRootConfigProvider } from '../config-provider';
import { DropdownMenuContext } from '../dropdown/DropdownMenuContext';
import './style';
const ContextMenu = props => {
  const {
    position,
    onClose,
    overlay
  } = props;
  const ref = React.useRef(null);
  const containerRef = React.useRef(null);
  const cssScope = useCssScope('context-menu');
  // 点击非menu区域后关闭
  useComplementEvent('click', {
    listener: React.useCallback(() => {
      onClose && onClose();
    }, [onClose]),
    elements: React.useMemo(() => [ref], [])
  });
  useLayoutEffect(() => {
    if (containerRef.current && ref.current) {
      // const Rect;
      const [left, top] = position;
      const result = {
        left,
        top
      };
      const el = ref.current;
      const cel = containerRef.current;
      const height = el.clientHeight;
      const width = el.clientWidth;

      // console.log('el', { left, top, height, width });
      // console.log('cel', { height: cel.clientHeight, width: cel.clientWidth });

      if (height + top > cel.clientHeight) {
        result.top = result.top - height;
        if (result.top < 0) {
          result.top = 0;
        }
      }
      if (width + left > cel.clientWidth) {
        result.left = left - width;
        if (result.left < 0) {
          result.left = 0;
        }
      }
      ref.current.style.left = `${result.left}px`;
      ref.current.style.top = `${result.top}px`;
    }
  }, [position]);
  const overlayElement = React.useMemo(() => isFunction(overlay) ? overlay() : overlay, [overlay]);
  return (
    /*#__PURE__*/
    // 为了告诉 menu 这是一个 dropdown
    React.createElement(DropdownMenuContext.Provider, {
      value: React.useMemo(() => ({}), [])
    }, /*#__PURE__*/React.createElement("div", {
      ref: containerRef,
      className: cssScope('__wrap'),
      onContextMenu: e => e.stopPropagation()
    }, /*#__PURE__*/React.createElement("div", {
      ref: ref,
      className: cssScope('__overlay'),
      onClick: onClose
    }, overlayElement)))
  );
};
export const contextMenu = {
  open: option => {
    const {
      menu,
      position,
      getPopupContainer,
      onClose
    } = option;

    // const uuid = `ud-confirm-instance-${new Date().getTime()}`;

    const portalEl = getPopupContainer ? getPopupContainer() : document.body;
    const container = document.createElement('div');
    portalEl.appendChild(container);
    ReactDOM.render(/*#__PURE__*/React.createElement(CrossRootConfigProvider, null, /*#__PURE__*/React.createElement(ContextMenu, {
      overlay: menu,
      position: position,
      onClose: () => {
        const unmountResult = ReactDOM.unmountComponentAtNode(container);
        if (unmountResult && container.parentNode) {
          container.parentNode.removeChild(container);
          onClose && onClose();
        }
      }
    })), container);
  }
};
