// @universe-design/react/lib/util/dom.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.contains = contains;
exports.eventContains = eventContains;
exports.getActiveElement = getActiveElement;
exports.getOwnerDocument = getOwnerDocument;
exports.getRelatedTarget = getRelatedTarget;
exports.isHorizontalOverflowed = isHorizontalOverflowed;
exports.isInputElement = isInputElement;
exports.isOverflowed = isOverflowed;
exports.isVerticalOverflowed = void 0;
var _reactUtilities = require("@universe-design/react-utilities");
function getOwnerDocument(node) {
  var _node$ownerDocument;
  return (0, _reactUtilities.isElement)(node) ? (_node$ownerDocument = node.ownerDocument) !== null && _node$ownerDocument !== void 0 ? _node$ownerDocument : document : document;
}
function isInputElement(el) {
  return (0, _reactUtilities.isHTMLElement)(el) && el.tagName.toLowerCase() === 'input' && 'select' in el;
}
function contains(parent, child) {
  if (!parent) {
    return false;
  }
  if (parent === child) {
    return true;
  }
  if ('contains' in parent) {
    return parent.contains(child);
  }
  return false;
}
function eventContains(parent, ev) {
  if (!parent) {
    return false;
  }
  const {
    target
  } = ev;
  if (!(0, _reactUtilities.isElement)(target)) {
    return false;
  }
  if (ev.composedPath) {
    // 处理 shadowRoot 情况
    return ev.composedPath().indexOf(parent) !== -1;
  }
  return contains(parent, target);
}
function getActiveElement(node) {
  const doc = getOwnerDocument(node);
  return doc === null || doc === void 0 ? void 0 : doc.activeElement;
}
function getRelatedTarget(event) {
  var _event$target, _event$relatedTarget;
  const target = (_event$target = event.target) !== null && _event$target !== void 0 ? _event$target : event.currentTarget;
  const activeElement = getActiveElement(target);
  return (_event$relatedTarget = event.relatedTarget) !== null && _event$relatedTarget !== void 0 ? _event$relatedTarget : activeElement;
}
function isHorizontalOverflowed(element) {
  // 这个函数用于判断单行文本是否溢出，只需要判断 x 方向即可
  return element.scrollWidth > element.clientWidth;
}
const isVerticalOverflowed = ele =>
// 这个函数用于判断多行文本是否溢出，只需要考虑 y 方向
ele.clientHeight < ele.scrollHeight;
exports.isVerticalOverflowed = isVerticalOverflowed;
function isOverflowed(element) {
  // 默认的溢出判断逻辑，包含 x 和 y 两个方向，虽然在用户 line-height 设置有问题的时候可能会有一些误判，但是我们认为用户 line-height 小于内容高度是一种「错误」。
  return isHorizontalOverflowed(element) || isVerticalOverflowed(element);
}
