// @universe-design/react/lib/util/size.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SizeEnum = exports.DefaultSizeEnum = exports.DefaultExtraSizeEnum = void 0;
exports.getDefaultSizeAbbr = getDefaultSizeAbbr;
Object.defineProperty(exports, "getSizeAbbr", {
  enumerable: true,
  get: function () {
    return _reactUtilities.getSizeAbbr;
  }
});
exports.isValidDefaultSize = isValidDefaultSize;
exports.isValidSize = isValidSize;
exports.sizeMap = void 0;
var _reactUtilities = require("@universe-design/react-utilities");
const SizeEnum = exports.SizeEnum = ['xl', 'lg', 'md', 'sm', 'xs'];
/**
 * @public
 */
const DefaultSizeEnum = exports.DefaultSizeEnum = ['large', 'medium', 'small'];
const DefaultExtraSizeEnum = exports.DefaultExtraSizeEnum = ['extra-large', 'extra-small'
// 'default',
];
const sizeMap = exports.sizeMap = {
  'extra-large': 'xl',
  large: 'lg',
  medium: 'md',
  // default: 'md',
  small: 'sm',
  'extra-small': 'xs'
};
function isValidDefaultSize(sizeName) {
  return DefaultSizeEnum.includes(sizeName);
}
function isValidSize(sizeName) {
  return DefaultSizeEnum.includes(sizeName) || DefaultExtraSizeEnum.includes(sizeName);
}
function getDefaultSizeAbbr(sizeName, defaultName) {
  if (isValidDefaultSize(sizeName)) {
    return sizeMap[sizeName];
  }
  if (defaultName && isValidDefaultSize(sizeName)) {
    return sizeMap[defaultName];
  }
  return '';
}
