// @universe-design/react/lib/util/tabbable.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.hasDisplayNone = void 0;
exports.hasFocusWithin = hasFocusWithin;
exports.hasTabIndex = exports.hasNegativeTabIndex = void 0;
exports.isActiveElement = isActiveElement;
exports.isContentEditable = isContentEditable;
exports.isDisabled = isDisabled;
exports.isFocusable = isFocusable;
exports.isHidden = isHidden;
exports.isTabbable = isTabbable;
var _reactUtilities = require("@universe-design/react-utilities");
var _dom = require("./dom");
const hasDisplayNone = element => window.getComputedStyle(element).display === 'none';
exports.hasDisplayNone = hasDisplayNone;
const hasTabIndex = element => element.hasAttribute('tabindex');
exports.hasTabIndex = hasTabIndex;
const hasNegativeTabIndex = element => hasTabIndex(element) && element.tabIndex === -1;
exports.hasNegativeTabIndex = hasNegativeTabIndex;
function isDisabled(element) {
  return Boolean(element.getAttribute('disabled')) === true || Boolean(element.getAttribute('aria-disabled')) === true;
}
function isActiveElement(element) {
  const doc = (0, _reactUtilities.isHTMLElement)(element) ? (0, _dom.getOwnerDocument)(element) : document;
  return doc.activeElement === element;
}
function hasFocusWithin(element) {
  if (!document.activeElement) {
    return false;
  }
  return element.contains(document.activeElement);
}
function isHidden(element) {
  if (element.parentElement && isHidden(element.parentElement)) {
    return true;
  }
  return element.hidden;
}
function isContentEditable(element) {
  const value = element.getAttribute('contenteditable');
  return value !== 'false' && value !== null;
}
function isFocusable(element) {
  if (!(0, _reactUtilities.isHTMLElement)(element) || isHidden(element) || isDisabled(element)) {
    return false;
  }
  const {
    localName
  } = element;
  if (['input', 'select', 'textarea', 'button'].indexOf(localName) >= 0) {
    return true;
  }
  const others = {
    a: () => element.hasAttribute('href'),
    audio: () => element.hasAttribute('controls'),
    video: () => element.hasAttribute('controls')
  };
  if (localName in others) {
    return others[localName]();
  }
  if (isContentEditable(element)) {
    return true;
  }
  return hasTabIndex(element);
}
function isTabbable(element) {
  if (!element) {
    return false;
  }
  return (0, _reactUtilities.isHTMLElement)(element) && isFocusable(element) && !hasNegativeTabIndex(element);
}
