// @universe-design/rc-table/lib/Body/ExpandedRow.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _typeof = require("@babel/runtime/helpers/typeof").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var React = _interopRequireWildcard(require("react"));
var _Cell = _interopRequireDefault(require("../Cell"));
var _TableContext = require("../context/TableContext");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function ExpandedRow(_ref) {
  var prefixCls = _ref.prefixCls,
    children = _ref.children,
    Component = _ref.component,
    cellComponent = _ref.cellComponent,
    fixHeader = _ref.fixHeader,
    fixColumn = _ref.fixColumn,
    horizonScroll = _ref.horizonScroll,
    className = _ref.className,
    expanded = _ref.expanded,
    componentWidth = _ref.componentWidth,
    colSpan = _ref.colSpan;
  var scrollbarColumnWidth = (0, _TableContext.useTableContextSelector)(function (ctx) {
    return ctx.scrollbarColumnWidth;
  });
  // Cache render node
  return React.useMemo(function () {
    var contentNode = children;
    if (fixColumn) {
      var width = componentWidth - scrollbarColumnWidth;
      contentNode = /*#__PURE__*/React.createElement("div", {
        style: {
          width: width,
          position: 'sticky',
          left: 0,
          overflow: 'hidden',
          visibility: width > 0 ? undefined : 'hidden' // 防止从宽度为0 到得到宽度时，出现闪烁
        },
        className: "".concat(prefixCls, "-expanded-row-fixed")
      }, contentNode);
    }
    return /*#__PURE__*/React.createElement(Component, {
      className: className,
      style: {
        display: expanded ? null : 'none'
      }
    }, /*#__PURE__*/React.createElement(_Cell.default, {
      component: cellComponent,
      prefixCls: prefixCls,
      colSpan: colSpan
    }, contentNode));
  }, [children, Component, fixHeader, horizonScroll, className, expanded, componentWidth, colSpan, scrollbarColumnWidth]);
}
var _default = ExpandedRow;
exports.default = _default;
