// @universe-design/rc-table/lib/Body/GroupRow.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _typeof = require("@babel/runtime/helpers/typeof").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GroupRow = GroupRow;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _Cell = _interopRequireDefault(require("../Cell"));
var _TableContext = require("../context/TableContext");
var _classnames = _interopRequireDefault(require("classnames"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function GroupRow(props) {
  var groupKey = props.groupKey,
    groupRowRenderer = props.groupRowRenderer,
    Component = props.component,
    cellComponent = props.cellComponent,
    colSpan = props.colSpan,
    prefixCls = props.prefixCls,
    triggerExpandGroup = props.triggerExpandGroup,
    onGroupRow = props.onGroupRow,
    componentWidth = props.componentWidth,
    horizonScroll = props.horizonScroll,
    expanded = props.expanded;
  var scrollbarColumnWidth = (0, _TableContext.useTableContextSelector)(function (ctx) {
    return ctx.scrollbarColumnWidth;
  });
  // if horizontalScroll we should keep it fixed and set it width equal to table width
  var contentNode = React.useMemo(function () {
    var renderGroupRowTitle = function renderGroupRowTitle() {
      if (typeof groupRowRenderer === 'function') {
        return groupRowRenderer(groupKey, expanded);
      }
      return groupKey;
    };
    return /*#__PURE__*/React.createElement("div", {
      style: {
        width: horizonScroll ? componentWidth - scrollbarColumnWidth : undefined,
        position: 'sticky',
        left: 0
      },
      className: "".concat(prefixCls, "-group-row-content")
    }, renderGroupRowTitle());
  }, [componentWidth, scrollbarColumnWidth, groupKey, groupRowRenderer, expanded, horizonScroll, prefixCls]);
  // Event
  var additionalProps;
  if (onGroupRow) {
    additionalProps = onGroupRow(groupKey);
  }
  var onClick = function onClick(ev) {
    triggerExpandGroup(groupKey);
    if (additionalProps && additionalProps.onClick) {
      var _additionalProps;
      for (var _len = arguments.length, args = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
        args[_key - 1] = arguments[_key];
      }
      (_additionalProps = additionalProps).onClick.apply(_additionalProps, [ev].concat(args));
    }
  };
  var mergedClassName = additionalProps && additionalProps.className ? (0, _classnames.default)(additionalProps.className, "".concat(prefixCls, "-group-row")) : "".concat(prefixCls, "-group-row");
  return /*#__PURE__*/React.createElement(Component, (0, _extends2.default)({}, additionalProps, {
    className: mergedClassName,
    onClick: onClick
  }), /*#__PURE__*/React.createElement(_Cell.default, {
    component: cellComponent,
    prefixCls: prefixCls,
    className: "".concat(prefixCls, "-group-row-cell"),
    colSpan: colSpan
  }, contentNode));
}
