// @universe-design/rc-table/lib/Body/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _typeof = require("@babel/runtime/helpers/typeof").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var React = _interopRequireWildcard(require("react"));
var _BodyRow = _interopRequireDefault(require("./BodyRow"));
var _TableContext = require("../context/TableContext");
var _ExpandedRow = _interopRequireDefault(require("./ExpandedRow"));
var _BodyContext = _interopRequireDefault(require("../context/BodyContext"));
var _valueUtil = require("../utils/valueUtil");
var _ResizeContext = _interopRequireDefault(require("../context/ResizeContext"));
var _MeasureCell = _interopRequireDefault(require("./MeasureCell"));
var _GroupRow = require("./GroupRow");
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function Body(_ref) {
  var data = _ref.data,
    getRowKey = _ref.getRowKey,
    measureColumnWidth = _ref.measureColumnWidth,
    expandedKeys = _ref.expandedKeys,
    onRow = _ref.onRow,
    rowExpandable = _ref.rowExpandable,
    emptyNode = _ref.emptyNode,
    childrenColumnName = _ref.childrenColumnName,
    groupsMap = _ref.groupsMap,
    groupRowRenderer = _ref.groupRowRenderer,
    expandedGroupKeySet = _ref.expandedGroupKeySet,
    triggerExpandGroup = _ref.triggerExpandGroup,
    onGroupRow = _ref.onGroupRow,
    getRecordByKey = _ref.getRecordByKey;
  var _React$useContext = React.useContext(_ResizeContext.default),
    onColumnResize = _React$useContext.onColumnResize;
  var prefixCls = (0, _TableContext.useTableContextSelector)(function (ctx) {
    return ctx.prefixCls;
  });
  var getComponent = (0, _TableContext.useTableContextSelector)(function (ctx) {
    return ctx.getComponent;
  });
  var _React$useContext2 = React.useContext(_BodyContext.default),
    fixHeader = _React$useContext2.fixHeader,
    horizonScroll = _React$useContext2.horizonScroll,
    flattenColumns = _React$useContext2.flattenColumns,
    componentWidth = _React$useContext2.componentWidth;
  var WrapperComponent = getComponent(['body', 'wrapper'], 'tbody');
  var trComponent = getComponent(['body', 'row'], 'tr');
  var tdComponent = getComponent(['body', 'cell'], 'td');
  return React.useMemo(function () {
    var columnsKey = (0, _valueUtil.getColumnsKey)(flattenColumns);
    // data row: lines of record
    var renderRows = function renderRows(rowsData) {
      var rows;
      if (rowsData.length) {
        rows = rowsData.map(function (record, index) {
          var key = getRowKey(record, index);
          // one record row and expanded row (if expandable)
          return /*#__PURE__*/React.createElement(_BodyRow.default, {
            key: key,
            rowKey: key,
            record: record,
            recordKey: key,
            index: index,
            rowComponent: trComponent,
            cellComponent: tdComponent,
            expandedKeys: expandedKeys,
            onRow: onRow,
            getRowKey: getRowKey,
            rowExpandable: rowExpandable,
            childrenColumnName: childrenColumnName
          });
        });
      } else {
        rows = /*#__PURE__*/React.createElement(_ExpandedRow.default, {
          expanded: true,
          className: "".concat(prefixCls, "-placeholder"),
          prefixCls: prefixCls,
          fixHeader: fixHeader,
          fixColumn: horizonScroll,
          horizonScroll: horizonScroll,
          component: trComponent,
          componentWidth: componentWidth,
          cellComponent: tdComponent,
          colSpan: flattenColumns.length
        }, emptyNode);
      }
      return rows;
    };
    // all table rows, contains group title、data row、expanded row
    var renderTableRows = function renderTableRows() {
      var groupKeys = (0, _toConsumableArray2.default)(groupsMap.keys());
      // grouped data
      if (!!groupKeys.length) {
        var groups = groupKeys.map(function (key) {
          var groupExpanded = expandedGroupKeySet.has(key);
          // group title row
          var renderGroupRow = function renderGroupRow() {
            return /*#__PURE__*/React.createElement(_GroupRow.GroupRow, {
              groupKey: key,
              prefixCls: prefixCls,
              groupRowRenderer: groupRowRenderer,
              triggerExpandGroup: triggerExpandGroup,
              onGroupRow: onGroupRow,
              component: trComponent,
              cellComponent: tdComponent,
              colSpan: flattenColumns.length,
              componentWidth: componentWidth,
              horizonScroll: horizonScroll,
              expanded: groupExpanded
            });
          };
          //group data rows
          var renderGroupedDataRows = function renderGroupedDataRows() {
            var groupDataKeys = groupsMap.get(key);
            var groupData = groupDataKeys.map(function (k) {
              return getRecordByKey(k);
            });
            return renderRows(groupData);
          };
          return /*#__PURE__*/React.createElement(React.Fragment, {
            key: key
          }, renderGroupRow(), groupExpanded && renderGroupedDataRows());
        });
        return groups;
      } else {
        return renderRows(data);
      }
    };
    return /*#__PURE__*/React.createElement(WrapperComponent, {
      className: "".concat(prefixCls, "-tbody")
    }, measureColumnWidth && /*#__PURE__*/React.createElement("tr", {
      "aria-hidden": "true",
      className: "".concat(prefixCls, "-measure-row"),
      style: {
        height: 0,
        fontSize: 0
      }
    }, columnsKey.map(function (columnKey) {
      return /*#__PURE__*/React.createElement(_MeasureCell.default, {
        key: columnKey,
        columnKey: columnKey,
        onColumnResize: onColumnResize
      });
    })), renderTableRows());
  }, [data, prefixCls, emptyNode, expandedGroupKeySet, expandedKeys, flattenColumns, getRowKey, measureColumnWidth, onRow, componentWidth, WrapperComponent, childrenColumnName, fixHeader, getRecordByKey, groupRowRenderer, groupsMap, horizonScroll, onColumnResize, onGroupRow, rowExpandable, tdComponent, trComponent, triggerExpandGroup]);
}
var MemoBody = /*#__PURE__*/React.memo(Body);
MemoBody.displayName = 'Body';
var _default = MemoBody;
exports.default = _default;
