// @universe-design/react-ai/lib-esm/card/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React, { useState, useMemo, forwardRef } from 'react';
import { composeRefs } from '@universe-design/react-utilities';
import { useResizeObserver } from '@universe-design/react';
import clsx from 'clsx';
import './styles/index.less';

/**
 * @public
 */

/**
 * @public
 */
export const AICard = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    children,
    className,
    processing = false,
    shadowed = true,
    noHoverState = false,
    ...restProps
  } = props;
  // aiCard 的尺寸
  const [{
    width,
    height
  }, setAiCardSize] = useState({
    width: 0,
    height: 0
  });

  // 获取 aiCard 的尺寸
  const aiCardRef = useResizeObserver(entry => {
    const {
      contentRect: {
        width: _width,
        height: _height
      }
    } = entry;
    setAiCardSize({
      width: _width,
      height: _height
    });
  });

  // 求容器对角线的长度
  const maskR = useMemo(() => Math.hypot(width, height), [height, width]);
  const mergedRef = useMemo(() => composeRefs(aiCardRef, ref), [aiCardRef, ref]);
  return /*#__PURE__*/React.createElement("div", _extends({
    ref: mergedRef,
    className: clsx('udx__ai__card', {
      ['udx__ai__card--default']: !processing,
      ['udx__ai__card--noshadows']: !shadowed,
      ['udx__ai__card--no-hover-state']: noHoverState
    }, className)
  }, restProps), processing && /*#__PURE__*/React.createElement("div", {
    className: "udx__ai__card__mask",
    style: {
      width: maskR,
      height: maskR,
      borderRadius: '50%',
      top: -(maskR - height) / 2,
      left: -(maskR - width) / 2
    }
  }), /*#__PURE__*/React.createElement("div", {
    className: clsx('udx__ai__card__content', processing && 'udx__ai__card__content--processing')
  }, children));
});
