// @universe-design/react/lib/tabs/ScrollableTabsHeader.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ScrollableTabsHeader = void 0;
var _u125g0x = _interopRequireDefault(require("./../.stylex/u125g0x8"));
var _ug24l6d = _interopRequireDefault(require("./../.stylex/ug24l6d"));
var _u1qr8jow = _interopRequireDefault(require("./../.stylex/u1qr8jow"));
var _u9vy1v = _interopRequireDefault(require("./../.stylex/u9vy1v2"));
var _u1no85ao = _interopRequireDefault(require("./../.stylex/u1no85ao"));
var _u1xegxa = _interopRequireDefault(require("./../.stylex/u1xegxa0"));
var _u5qwpn = _interopRequireDefault(require("./../.stylex/u5qwpn8"));
var _u1l27p0d = _interopRequireDefault(require("./../.stylex/u1l27p0d"));
var _ujik7se = _interopRequireDefault(require("./../.stylex/ujik7se"));
var _u1yoqygm = _interopRequireDefault(require("./../.stylex/u1yoqygm"));
var _u1anwibm = _interopRequireDefault(require("./../.stylex/u1anwibm"));
var _u1zgj7v = _interopRequireDefault(require("./../.stylex/u1zgj7v"));
var _u1w1hd4l = _interopRequireDefault(require("./../.stylex/u1w1hd4l"));
var _u1aoskr = _interopRequireDefault(require("./../.stylex/u1aoskr8"));
var _ubm18w = _interopRequireDefault(require("./../.stylex/ubm18w5"));
var _u1klsf9c = _interopRequireDefault(require("./../.stylex/u1klsf9c"));
var _uua3xia = _interopRequireDefault(require("./../.stylex/uua3xia"));
var _u16zaoyg = _interopRequireDefault(require("./../.stylex/u16zaoyg"));
var _u19z4mkn = _interopRequireDefault(require("./../.stylex/u19z4mkn"));
var _u1pid3wf = _interopRequireDefault(require("./../.stylex/u1pid3wf"));
var _ucgpi4a = _interopRequireDefault(require("./../.stylex/ucgpi4a"));
var _upyoa5l = _interopRequireDefault(require("./../.stylex/upyoa5l"));
var _u102c4m = _interopRequireDefault(require("./../.stylex/u102c4m8"));
var _u1mp4r5t = _interopRequireDefault(require("./../.stylex/u1mp4r5t"));
var _uc4om0q = _interopRequireDefault(require("./../.stylex/uc4om0q"));
var _urmav4p = _interopRequireDefault(require("./../.stylex/urmav4p"));
var _u99gbyh = _interopRequireDefault(require("./../.stylex/u99gbyh"));
var _uldrd5o = _interopRequireDefault(require("./../.stylex/uldrd5o"));
var _u13d3rh = _interopRequireDefault(require("./../.stylex/u13d3rh8"));
var _u1vjduc = _interopRequireDefault(require("./../.stylex/u1vjduc2"));
var _u14cxjev = _interopRequireDefault(require("./../.stylex/u14cxjev"));
var _ukchsq = _interopRequireDefault(require("./../.stylex/ukchsq7"));
var _ug4hjch = _interopRequireDefault(require("./../.stylex/ug4hjch"));
var _ubyhozq = _interopRequireDefault(require("./../.stylex/ubyhozq"));
var _uevmt6d = _interopRequireDefault(require("./../.stylex/uevmt6d"));
var _u283mfx = _interopRequireDefault(require("./../.stylex/u283mfx"));
var _u12o6wmt = _interopRequireDefault(require("./../.stylex/u12o6wmt"));
var _u1u0vkcg = _interopRequireDefault(require("./../.stylex/u1u0vkcg"));
var _u5jxv = _interopRequireDefault(require("./../.stylex/u5jxv71"));
var _u16ye = _interopRequireDefault(require("./../.stylex/u16ye362"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _react = require("@universe-design/stylex/react");
var _react2 = _interopRequireDefault(require("react"));
var _Tab = require("./Tab");
var _TabContext = require("./TabContext");
var _TabsContext = require("./TabsContext");
var _LeftOutlined = _interopRequireDefault(require("@universe-design/icons-react/LeftOutlined"));
var _RightOutlined = _interopRequireDefault(require("@universe-design/icons-react/RightOutlined"));
var _TabInk = require("./TabInk");
var _reactUtilities = require("@universe-design/react-utilities");
var _forwardRef = require("../util/forwardRef");
var _button = require("../button");
var _sizing_behavior = require("./sizing_behavior");
//.ulvyvdh{flex-wrap:nowrap;}
//.u1jhi6b8{flex-grow:0;}
//.u1oy3dpc{overflow-x:auto;}
//.uo9lwiw{scrollbar-width:none;}
//.um5o6js ::-webkit-scrollbar{height:0px;}
//.uqc6z8f:after{content:"";}
//.u11ef69:after{position:absolute;}
//.uqevjk0:after{pointer-events:none;}
//.u19zsqug:after{z-index:2;}
//.u1oarbwq:after{right:0px;}
//.u1yuyem3:after{top:0px;}
//.u1soztmy:after{bottom:6px;}
//.us0klr3:after{width:var(--scroll-mask-width);}
//.u2df6js:after{display:block;}
//.u16igepq:after{background:@udtoken-component-bg-right-mask;}
//.uuzzvh5:before{content:"";}
//.uxf9f1y:before{position:absolute;}
//.urcipv9:before{pointer-events:none;}
//.u13tv88k:before{z-index:2;}
//.u8g0ba1:before{left:0px;}
//.urdto5n:before{top:0px;}
//.u14i0ki9:before{bottom:6px;}
//.ukapjos:before{width:var(--scroll-mask-width);}
//.uxv2tmp:before{display:block;}
//.u1801gqn:before{background:@udtoken-component-bg-left-mask;}
//.u10tiqix{min-width:0px;}
//.uqerorx{flex-grow:1;}
//.u1neuvcm{flex-shrink:1;}
//.uly5x3f{width:100%;}
//.udrzuqr{cursor:not-allowed;}
//.u1qs8jb8{color:@icon-disabled;}
//.u22iagw{display:flex;}
//.u1ub7u0d{align-self:stretch;}
//.u122n59{align-items:center;}
//.u64fuq3{width:20px;}
//.u1ugzwwg{font-size:12px;}
//.u1k6fduh{cursor:pointer;}
//.u1bu1o99{color:@icon-n1;}
//.u10pi13n{position:relative;}
//.u17gl07{background:@udtoken-component-bg-right-mask;}
/* eslint-disable react/no-unknown-property */
/** @jsxRuntime classic */
/** @jsx jsx */
/** @jsxFrag React.Fragment */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const defaultScrollOption = {
  // 外层有滚动容器的话，外层不跟随滚动
  block: 'nearest',
  behavior: 'smooth'
};
const ScrollableTabsHeader = exports.ScrollableTabsHeader = (0, _forwardRef.forwardHTMLRef)((_ref, ref) => {
  let {
    inkOffsetY
  } = _ref;
  const {
    rootCssScope,
    panes,
    scrollButtonGutter,
    tabElementMapRef,
    type,
    rootTabsSizingBehavior
  } = (0, _TabsContext.useTabsContext)();
  const [scrollButtonVisible, setScrollButtonVisible] = _react2.default.useState(false);
  const [scrollButtonDisabled, setScrollButtonDisabled] = _react2.default.useState([true, false]);
  const scrollMaskWidth = type === 'line' ? 22 : 0;
  const resizeHandler = () => {
    if (wrapperRef.current) {
      const {
        scrollWidth,
        clientWidth
      } = wrapperRef.current;
      setScrollButtonVisible(scrollWidth - clientWidth > 1);
      checkScrollButtonStatus();
    }
  };
  // 容器尺寸变化（往往是受到挤压或者缩放）
  const wrapperRef = (0, _reactUtilities.useResizeObserver)(resizeHandler);
  // 列表尺寸变化（往往是数据变化）
  const innerListRef = (0, _reactUtilities.useResizeObserver)(resizeHandler);
  const composedRef = (0, _reactUtilities.composeRefs)(wrapperRef, ref);

  /**
   * 当 tabBar 滚动时，调整 ScrollButton 的禁用态
   */
  const checkScrollButtonStatus = _react2.default.useCallback(() => {
    if (!wrapperRef.current) {
      return;
    }
    const leftButtonDisabled = wrapperRef.current.scrollLeft === 0;
    const rightButtonDisabled = Math.round(wrapperRef.current.scrollLeft + wrapperRef.current.clientWidth) >= Math.round(wrapperRef.current.scrollWidth);
    setScrollButtonDisabled([leftButtonDisabled, rightButtonDisabled]);
  }, [wrapperRef]);

  /**
   * 从左向右滚动
   */
  const gotoNextPage = _react2.default.useCallback(() => {
    var _wrapperRef$current;
    const containerRight = ((_wrapperRef$current = wrapperRef.current) === null || _wrapperRef$current === void 0 ? void 0 : _wrapperRef$current.getBoundingClientRect().right) || 0;
    /**
     * 找到第一个未显示完全的 tab
     */
    for (const pane of panes) {
      var _tabElementMapRef$cur;
      const tab = (_tabElementMapRef$cur = tabElementMapRef.current) === null || _tabElementMapRef$cur === void 0 ? void 0 : _tabElementMapRef$cur.get(pane.key);
      if (!tab) {
        continue;
      }
      const {
        right: tabRight
      } = tab.getBoundingClientRect();
      if (tabRight > containerRight) {
        tab.scrollIntoView({
          ...defaultScrollOption,
          // 让选项滚动到最左侧，而不是仅仅滚动到可见区域
          inline: 'start'
        });
        return;
      }
    }
  }, [panes, tabElementMapRef, wrapperRef]);

  /**
   * 从右向左滚动
   */
  const gotoPrePage = _react2.default.useCallback(() => {
    var _wrapperRef$current2;
    const containerLeft = ((_wrapperRef$current2 = wrapperRef.current) === null || _wrapperRef$current2 === void 0 ? void 0 : _wrapperRef$current2.getBoundingClientRect().left) || 0;
    /**
     * 找到第一个未显示完全的 tab
     */
    for (const pane of [...panes].reverse()) {
      var _tabElementMapRef$cur2;
      const tab = (_tabElementMapRef$cur2 = tabElementMapRef.current) === null || _tabElementMapRef$cur2 === void 0 ? void 0 : _tabElementMapRef$cur2.get(pane.key);
      if (!tab) {
        continue;
      }
      const {
        left: tabLeft
      } = tab.getBoundingClientRect();
      if (tabLeft < containerLeft) {
        tab.scrollIntoView({
          ...defaultScrollOption,
          // 让选项滚动到最右侧，而不是仅仅滚动到可见区域
          inline: 'end'
        });
        return;
      }
    }
  }, [panes, tabElementMapRef, wrapperRef]);
  const getScrollButtonSx = position => [type !== 'capsule' && {
    "props": {
      "mc9l5x": _ukchsq.default,
      "qb2dma": _ug4hjch.default,
      "Bt984gj": _ubyhozq.default,
      "a9b677": _uevmt6d.default,
      "Be2twd7": _u283mfx.default,
      "Bceei9c": _u12o6wmt.default,
      "sj55zd": _u1u0vkcg.default,
      "qhf8xq": _u5jxv.default,
      "ayd6f0": _u16ye.default
    },
    "alias": rootCssScope('__scroll-button')
  }, (position === 'left' ? scrollButtonDisabled[0] : scrollButtonDisabled[1]) && {
    "props": {
      "Bceei9c": _u1vjduc.default,
      "sj55zd": _u14cxjev.default
    }
  }];

  /**
   * 选中 tab,将 tabBar 滚动到可视区域内
   */
  const scrollToSelectedTab = _react2.default.useCallback(key => {
    var _tabElementMapRef$cur3;
    const tab = (_tabElementMapRef$cur3 = tabElementMapRef.current) === null || _tabElementMapRef$cur3 === void 0 ? void 0 : _tabElementMapRef$cur3.get(key);
    const wrapperEle = wrapperRef.current;
    if (!tab || !wrapperEle) {
      return;
    }
    const {
      right: containerRight,
      left: containerLeft
    } = wrapperEle.getBoundingClientRect();
    const {
      right: tabRight,
      left: tabLeft
    } = tab.getBoundingClientRect();

    // 在左侧
    if (tabLeft < containerLeft + scrollMaskWidth) {
      wrapperEle.scroll({
        // 此时可能出现负数，但浏览器能兼容
        left: tab.offsetLeft - scrollMaskWidth,
        ...defaultScrollOption
      });
      return;
    }

    // 在右侧
    if (tabRight > containerRight - scrollMaskWidth) {
      wrapperEle.scroll({
        left: wrapperEle.scrollLeft + (tabRight - containerRight) + scrollMaskWidth,
        ...defaultScrollOption
      });
    }
  }, [scrollMaskWidth, tabElementMapRef, wrapperRef]);
  return (0, _react.jsx)("div", {
    style: {
      columnGap: scrollButtonGutter
    },
    sx: [{
      "props": {
        "mc9l5x": _ukchsq.default,
        "Bt984gj": _ubyhozq.default,
        "a9b677": _u13d3rh.default
      }
    }]
  }, Boolean(scrollButtonVisible) && (0, _react.jsx)("div", {
    sx: getScrollButtonSx('left'),
    onClick: scrollButtonDisabled[0] ? undefined : gotoPrePage
  }, (0, _react.jsx)(_button.IconButton, {
    size: "extra-small",
    disabled: scrollButtonDisabled[0],
    icon: (0, _react.jsx)(_LeftOutlined.default, null)
  })), (0, _react.jsx)("div", {
    style: {
      '--scroll-mask-width': scrollMaskWidth + 'px'
    },
    sx: [{
      "props": {
        "qhf8xq": _u5jxv.default,
        "Bf4jedk": _urmav4p.default,
        "Bh6795r": _u99gbyh.default,
        "Bnnss6s": _uldrd5o.default
      }
    }, !scrollButtonDisabled[0] && {
      "props": {
        "rurcny": _u1klsf9c.default,
        "xx9plb": _uua3xia.default,
        "iyvzjs": _u16zaoyg.default,
        "Bab4cu6": _u19z4mkn.default,
        "Byque4d": _u1pid3wf.default,
        "Bf8kmfk": _ucgpi4a.default,
        "Bj2wrql": _upyoa5l.default,
        "Bdn98qo": _u102c4m.default,
        "pmf9yy": _u1mp4r5t.default,
        "B4zgs9e": _uc4om0q.default
      }
    }, !scrollButtonDisabled[1] && {
      "props": {
        "Bs6t6z0": _u1xegxa.default,
        "Hdbjpj": _u5qwpn.default,
        "Bmgsduf": _u1l27p0d.default,
        "Bunff0j": _ujik7se.default,
        "cnmfks": _u1yoqygm.default,
        "u6d25": _u1anwibm.default,
        "Bj55yzk": _u1zgj7v.default,
        "Biobvvw": _u1w1hd4l.default,
        "Fdvyjd": _u1aoskr.default,
        "Bpudnjh": _ubm18w.default
      }
    }]
  }, (0, _react.jsx)("div", {
    ref: composedRef,
    className: rootCssScope('__tab-bar-wrap'),
    onScroll: checkScrollButtonStatus,
    sx: [{
      "props": {
        "B68tc82": _u1qr8jow.default,
        "qhf8xq": _u5jxv.default,
        "B15if2v": _u9vy1v.default,
        "qgh3wz": _u1no85ao.default
      }
    }]
  }, (0, _react.jsx)(_TabInk.TabsInkBar, {
    offsetY: inkOffsetY
  }), (0, _react.jsx)("div", {
    ref: innerListRef,
    sx: [{
      "props": {
        "mc9l5x": _ukchsq.default,
        "Eh141a": _u125g0x.default,
        "qhf8xq": _u5jxv.default,
        "Bnnss6s": _uldrd5o.default,
        "Bh6795r": _ug24l6d.default
      }
    }]
  }, panes.map(pane => (0, _react.jsx)("div", {
    key: pane.key,
    style: (0, _sizing_behavior.resolveSizingBehavior)(rootTabsSizingBehavior)
  }, (0, _react.jsx)(_TabContext.TabProvider, {
    tabKey: pane.key
  }, (0, _react.jsx)(_Tab.Tab, {
    onClick: () => scrollToSelectedTab(pane.key)
  }))))))), Boolean(scrollButtonVisible) && (0, _react.jsx)("div", {
    sx: getScrollButtonSx('right'),
    onClick: scrollButtonDisabled[1] ? undefined : gotoNextPage
  }, (0, _react.jsx)(_button.IconButton, {
    size: "extra-small",
    disabled: scrollButtonDisabled[1],
    icon: (0, _react.jsx)(_RightOutlined.default, null)
  })));
});
