// @universe-design/react/lib/tabs/sx.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useTabsMoreStyles = exports.useTabStyles = exports.sxTabTypography = exports.sxTabSize = void 0;
var _u1773ck = _interopRequireDefault(require("./../.stylex/u1773ck5"));
var _upqf1qz = _interopRequireDefault(require("./../.stylex/upqf1qz"));
var _u1cxaokx = _interopRequireDefault(require("./../.stylex/u1cxaokx"));
var _uzx399i = _interopRequireDefault(require("./../.stylex/uzx399i"));
var _u8lm1xd = _interopRequireDefault(require("./../.stylex/u8lm1xd"));
var _u1fzbsm = _interopRequireDefault(require("./../.stylex/u1fzbsm8"));
var _u1x6icpj = _interopRequireDefault(require("./../.stylex/u1x6icpj"));
var _uoa5q = _interopRequireDefault(require("./../.stylex/uoa5q5"));
var _u1xrgxpq = _interopRequireDefault(require("./../.stylex/u1xrgxpq"));
var _uwlx52q = _interopRequireDefault(require("./../.stylex/uwlx52q"));
var _u10ismar = _interopRequireDefault(require("./../.stylex/u10ismar"));
var _u1g1nbrb = _interopRequireDefault(require("./../.stylex/u1g1nbrb"));
var _uia0x1d = _interopRequireDefault(require("./../.stylex/uia0x1d"));
var _uue = _interopRequireDefault(require("./../.stylex/uue9938"));
var _uf0hpal = _interopRequireDefault(require("./../.stylex/uf0hpal"));
var _ueh72r = _interopRequireDefault(require("./../.stylex/ueh72r3"));
var _u19b90lh = _interopRequireDefault(require("./../.stylex/u19b90lh"));
var _u1vk5ola = _interopRequireDefault(require("./../.stylex/u1vk5ola"));
var _u1n7c7ic = _interopRequireDefault(require("./../.stylex/u1n7c7ic"));
var _ujkvyav = _interopRequireDefault(require("./../.stylex/ujkvyav"));
var _uqbl7tl = _interopRequireDefault(require("./../.stylex/uqbl7tl"));
var _u19ubm5t = _interopRequireDefault(require("./../.stylex/u19ubm5t"));
var _umy11dx = _interopRequireDefault(require("./../.stylex/umy11dx"));
var _ucbcxg = _interopRequireDefault(require("./../.stylex/ucbcxg8"));
var _u1cd5c9u = _interopRequireDefault(require("./../.stylex/u1cd5c9u"));
var _uda9nr = _interopRequireDefault(require("./../.stylex/uda9nr0"));
var _uy9crqm = _interopRequireDefault(require("./../.stylex/uy9crqm"));
var _uptwv6n = _interopRequireDefault(require("./../.stylex/uptwv6n"));
var _u1cpcpb = _interopRequireDefault(require("./../.stylex/u1cpcpb5"));
var _u10qn3go = _interopRequireDefault(require("./../.stylex/u10qn3go"));
var _uflkg = _interopRequireDefault(require("./../.stylex/uflkg16"));
var _uvn5txf = _interopRequireDefault(require("./../.stylex/uvn5txf"));
var _us9j13g = _interopRequireDefault(require("./../.stylex/us9j13g"));
var _umt7kug = _interopRequireDefault(require("./../.stylex/umt7kug"));
var _uh3lwu = _interopRequireDefault(require("./../.stylex/uh3lwu4"));
var _ur = _interopRequireDefault(require("./../.stylex/ur61548"));
var _u1ymqkzg = _interopRequireDefault(require("./../.stylex/u1ymqkzg"));
var _u1w9jt9n = _interopRequireDefault(require("./../.stylex/u1w9jt9n"));
var _u1hdaclu = _interopRequireDefault(require("./../.stylex/u1hdaclu"));
var _ut7q8es = _interopRequireDefault(require("./../.stylex/ut7q8es"));
var _un6nsv = _interopRequireDefault(require("./../.stylex/un6nsv3"));
var _u15pdtm = _interopRequireDefault(require("./../.stylex/u15pdtm"));
var _u1y2b6fc = _interopRequireDefault(require("./../.stylex/u1y2b6fc"));
var _u1a72m5n = _interopRequireDefault(require("./../.stylex/u1a72m5n"));
var _uvnwla = _interopRequireDefault(require("./../.stylex/uvnwla"));
var _ujixoer = _interopRequireDefault(require("./../.stylex/ujixoer"));
var _u1i2nmr = _interopRequireDefault(require("./../.stylex/u1i2nmr7"));
var _ubr2xhg = _interopRequireDefault(require("./../.stylex/ubr2xhg"));
var _u1fdqv2r = _interopRequireDefault(require("./../.stylex/u1fdqv2r"));
var _u1oqvdpm = _interopRequireDefault(require("./../.stylex/u1oqvdpm"));
var _u6qvdto = _interopRequireDefault(require("./../.stylex/u6qvdto"));
var _ub7yml = _interopRequireDefault(require("./../.stylex/ub7yml4"));
var _u1kxgv9m = _interopRequireDefault(require("./../.stylex/u1kxgv9m"));
var _unbze3l = _interopRequireDefault(require("./../.stylex/unbze3l"));
var _u10d06an = _interopRequireDefault(require("./../.stylex/u10d06an"));
var _u1w6ojc = _interopRequireDefault(require("./../.stylex/u1w6ojc6"));
var _u5jxv = _interopRequireDefault(require("./../.stylex/u5jxv71"));
var _uhppl2f = _interopRequireDefault(require("./../.stylex/uhppl2f"));
var _u1e84ebu = _interopRequireDefault(require("./../.stylex/u1e84ebu"));
var _u8ckwbm = _interopRequireDefault(require("./../.stylex/u8ckwbm"));
var _upgelb = _interopRequireDefault(require("./../.stylex/upgelb9"));
var _u1vsjzop = _interopRequireDefault(require("./../.stylex/u1vsjzop"));
var _ujgl7ix = _interopRequireDefault(require("./../.stylex/ujgl7ix"));
var _ucil = _interopRequireDefault(require("./../.stylex/ucil735"));
var _u441rjq = _interopRequireDefault(require("./../.stylex/u441rjq"));
var _uaonily = _interopRequireDefault(require("./../.stylex/uaonily"));
var _um3isuf = _interopRequireDefault(require("./../.stylex/um3isuf"));
var _un7s8z = _interopRequireDefault(require("./../.stylex/un7s8z9"));
var _u143d86t = _interopRequireDefault(require("./../.stylex/u143d86t"));
var _u1ibfn = _interopRequireDefault(require("./../.stylex/u1ibfn25"));
var _urjm3ap = _interopRequireDefault(require("./../.stylex/urjm3ap"));
var _u1z1dmk = _interopRequireDefault(require("./../.stylex/u1z1dmk"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _stylex2 = require("../stylex");
var _react = require("@universe-design/stylex/react");
//.uvo9xlb{border-top-left-radius:@border-radius-md;}
//.ugd3wmz{border-top-right-radius:@border-radius-md;}
//.ueep33z{border-bottom-right-radius:@border-radius-md;}
//.u3qetca{border-bottom-left-radius:@border-radius-md;}
//.u1l2lv2i:hover{border-top-left-radius:@border-radius-md;}
//.u7l4e48:hover{border-top-right-radius:@border-radius-md;}
//.u1p09kk6:hover{border-bottom-right-radius:@border-radius-md;}
//.uhijqq1:hover{border-bottom-left-radius:@border-radius-md;}
//.u1ipgtue{background-color:@udtoken-component-outlined-bg;}
//.ub99si7{border-top-color:@line-border-component;}
//.u12908r0{border-right-color:@line-border-component;}
//.u1lgt5kb{border-bottom-color:@line-border-component;}
//.u3ps6bl{border-left-color:@line-border-component;}
//.u1es3o6n{color:@primary-content-default;}
//.urdqxp0:hover{background-color:var(--udtoken-tabs-view-bg-hover);}
//.u1rvp7gu{color:@text-caption;}
//.u3rkx2v{border-bottom-right-radius:@border-radius-sm;}
//.u17hi4zz{border-bottom-left-radius:@border-radius-sm;}
//.uyxjys6{border-top-color:@line-border-card;}
//.u1f4hk3w{border-right-color:@line-border-card;}
//.u14wfv0u{border-left-color:@line-border-card;}
//.u96dupq{border-top-left-radius:@border-radius-sm;}
//.u1msb61c{border-top-right-radius:@border-radius-sm;}
//.u1b829s6{background-color:@bg-base;}
//.u1p3nwhy{border-top-color:transparent;}
//.u11589ue{border-right-color:transparent;}
//.u1q5o8ev{border-bottom-color:transparent;}
//.u1pdflbu{border-left-color:transparent;}
//.u5ogflp{border-top-width:1px;}
//.u1hqa2wf{border-right-width:1px;}
//.u1f09k3d{border-bottom-width:1px;}
//.uinvdd3{border-left-width:1px;}
//.uzkkow9{border-top-style:solid;}
//.ucdblym{border-right-style:solid;}
//.ug706s2{border-bottom-style:solid;}
//.ujik90z{border-left-style:solid;}
//.u1eyxnol:hover:not(.u1eyxnol--disabled){color:@primary-content-default;}
//.u1kcqot9{padding-top:12px;}
//.upe6lb7{padding-bottom:12px;}
//.u4d9j23{justify-content:center;}
//.uxldao9{height:28px;}
//.u1sbtcvk{padding-top:5px;}
//.udghr9{padding-bottom:5px;}
//.u1xp5gbu{padding-top:6px;}
//.u1d7kygh{padding-bottom:6px;}
//.u1d2rq10{height:32px;}
//.u8ljn23{height:36px;}
//.ubhnoac{height:40px;}
//.u1g0x7ka{padding-top:0;}
//.uat0sn4{padding-right:20px;}
//.u1qch9an{padding-bottom:0;}
//.uekwl8i{padding-left:20px;}
//.u1lmfglv{margin-top:0px;}
//.uibjyge{margin-right:0px;}
//.u1abmfm4{margin-bottom:0px;}
//.u9yszdx{margin-left:0px;}
//.u10pi13n{position:relative;}
//.u19g0ac{z-index:1;}
//.up9bwmr{padding-top:8px;}
//.u150uoa4{padding-bottom:8px;}
//.uq1loh5{padding-top:2px;}
//.uutqtb8{padding-bottom:2px;}
//.u1phki43{margin-right:8px;}
//.u1khb0e9{padding-top:3px;}
//.u19lj068{padding-right:8px;}
//.u1jnq6q7{padding-bottom:3px;}
//.u177v4lu{padding-left:8px;}
//.u10ra9hq{padding-top:4px;}
//.u11qrl6u{padding-right:12px;}
//.u1y2xyjm{padding-bottom:4px;}
//.ujlbh76{padding-left:12px;}
//.u1x4n1mu{margin-right:12px;}
/**
 * Stylex CSS
 */
// Tab 的版面样式
const sxTabTypography = (size, isSelected) => {
  if (size === 'large') {
    if (isSelected) {
      return _stylex2.cssTypographyTitle4;
    }
    return _stylex2.cssTypographyTitle5;
  }
  if (size === 'medium') {
    if (isSelected) {
      return _stylex2.cssTypographyHeadline;
    }
    return _stylex2.cssTypographyBody0;
  }
  if (isSelected) {
    return _stylex2.cssTypographyCaption0;
  }
  return _stylex2.cssTypographyBody2;
};
exports.sxTabTypography = sxTabTypography;
const sxTabSize = (type, size) => {
  var _tabSizes$type, _tabSizes$type2;
  const tabSizes = {
    tag: {
      large: {
        "props": {
          "z8tnut": _un7s8z.default,
          "z189sj": _u143d86t.default,
          "Byoj8tv": _u1ibfn.default,
          "uwmqm3": _urjm3ap.default,
          "t21cq0": _u1z1dmk.default
        }
      },
      medium: {
        "props": {
          "z8tnut": _ucil.default,
          "z189sj": _u441rjq.default,
          "Byoj8tv": _uaonily.default,
          "uwmqm3": _um3isuf.default,
          "t21cq0": _u1z1dmk.default
        }
      },
      small: {
        "props": {
          "z8tnut": _upgelb.default,
          "z189sj": _u441rjq.default,
          "Byoj8tv": _u1vsjzop.default,
          "uwmqm3": _um3isuf.default,
          "t21cq0": _ujgl7ix.default
        }
      }
    },
    line: {
      small: {
        "props": {
          "z8tnut": _u1e84ebu.default,
          "Byoj8tv": _u8ckwbm.default
        }
      }
    },
    card: {
      base: {
        "props": {
          "B6of3ja": _u1kxgv9m.default,
          "t21cq0": _unbze3l.default,
          "jrapky": _u10d06an.default,
          "Frg6f3": _u1w6ojc.default,
          "qhf8xq": _u5jxv.default,
          "Bj3rh1h": _uhppl2f.default
        }
      },
      large: {
        "props": {
          "Bqenvij": _ubr2xhg.default,
          "z8tnut": _u1fdqv2r.default,
          "z189sj": _u1oqvdpm.default,
          "Byoj8tv": _u6qvdto.default,
          "uwmqm3": _ub7yml.default
        }
      },
      medium: {
        "props": {
          "Bqenvij": _u1i2nmr.default,
          "z8tnut": _u1fdqv2r.default,
          "z189sj": _u1oqvdpm.default,
          "Byoj8tv": _u6qvdto.default,
          "uwmqm3": _ub7yml.default
        }
      },
      small: {
        "props": {
          "Bqenvij": _ujixoer.default,
          "z8tnut": _u1fdqv2r.default,
          "z189sj": _u1oqvdpm.default,
          "Byoj8tv": _u6qvdto.default,
          "uwmqm3": _ub7yml.default
        }
      }
    },
    capsule: {
      large: {
        "props": {
          "Bqenvij": _u1i2nmr.default,
          "z8tnut": _u1a72m5n.default,
          "z189sj": _u143d86t.default,
          "Byoj8tv": _uvnwla.default,
          "uwmqm3": _urjm3ap.default
        }
      },
      medium: {
        "props": {
          "Bqenvij": _ujixoer.default,
          "z8tnut": _u15pdtm.default,
          "z189sj": _u143d86t.default,
          "Byoj8tv": _u1y2b6fc.default,
          "uwmqm3": _urjm3ap.default
        }
      },
      small: {
        "props": {
          "Bqenvij": _un6nsv.default,
          "z8tnut": _un7s8z.default,
          "z189sj": _u143d86t.default,
          "Byoj8tv": _u1ibfn.default,
          "uwmqm3": _urjm3ap.default
        }
      }
    }
  };
  return _stylex.default.compose({
    "props": {
      "z8tnut": _u1w9jt9n.default,
      "Byoj8tv": _u1hdaclu.default,
      "Brf1p80": _ut7q8es.default
    }
  }, (_tabSizes$type = tabSizes[type]) === null || _tabSizes$type === void 0 ? void 0 : _tabSizes$type.base, (_tabSizes$type2 = tabSizes[type]) === null || _tabSizes$type2 === void 0 ? void 0 : _tabSizes$type2[size]);
};
exports.sxTabSize = sxTabSize;
const useTabStyles = exports.useTabStyles = (0, _react.makeStylexStyles)(_ref => {
  let {
    prefix
  } = _ref;
  return {
    tabHover: {
      "props": {
        "lf9awn": _u1ymqkzg.default
      }
    },
    cardBase: {
      "props": {
        "B4j52fo": _u1cpcpb.default,
        "Bekrc4i": _u10qn3go.default,
        "Bn0qgzm": _uflkg.default,
        "ibv6hh": _uvn5txf.default,
        "icvyot": _us9j13g.default,
        "vrafjx": _umt7kug.default,
        "oivjwe": _uh3lwu.default,
        "wvpqe5": _ur.default
      }
    },
    cardDefault: {
      "props": {
        "g2u3we": _u1cd5c9u.default,
        "h3c5rm": _uda9nr.default,
        "B9xav0g": _uy9crqm.default,
        "zhjwy3": _uptwv6n.default
      },
      "alias": `${prefix}__tab--card-with-border`
    },
    cardActive: {
      "props": {
        "g2u3we": _u1n7c7ic.default,
        "h3c5rm": _ujkvyav.default,
        "B9xav0g": _uy9crqm.default,
        "zhjwy3": _uqbl7tl.default,
        "Btl43ni": _u19ubm5t.default,
        "B7oj6ja": _umy11dx.default,
        "De3pzq": _ucbcxg.default,
        "Bj3rh1h": _uhppl2f.default
      },
      "alias": `${prefix}__tab--card-with-border--active`
    },
    capsuleBase: {
      "props": {
        "sj55zd": _ueh72r.default,
        "B4j52fo": _u1cpcpb.default,
        "Bekrc4i": _u10qn3go.default,
        "Bn0qgzm": _uflkg.default,
        "ibv6hh": _uvn5txf.default,
        "icvyot": _us9j13g.default,
        "vrafjx": _umt7kug.default,
        "oivjwe": _uh3lwu.default,
        "wvpqe5": _ur.default,
        "Btl43ni": _u19ubm5t.default,
        "B7oj6ja": _umy11dx.default,
        "Bbmb7ep": _u19b90lh.default,
        "Beyfa6y": _u1vk5ola.default
      }
    },
    capsuleDefault: {
      "props": {
        "g2u3we": _u1cd5c9u.default,
        "h3c5rm": _uda9nr.default,
        "B9xav0g": _uy9crqm.default,
        "zhjwy3": _uptwv6n.default,
        "Jwef8y": _uf0hpal.default
      },
      "alias": `${prefix}__tab--capsule-with-border`
    },
    capsuleActive: {
      "props": {
        "De3pzq": _u1xrgxpq.default,
        "g2u3we": _uwlx52q.default,
        "h3c5rm": _u10ismar.default,
        "B9xav0g": _u1g1nbrb.default,
        "zhjwy3": _uia0x1d.default,
        "sj55zd": _uue.default
      },
      "alias": `${prefix}__tab--capsule-with-border--active`
    }
  };
});
const useTabsMoreStyles = exports.useTabsMoreStyles = (0, _react.makeStylexStyles)(_ref2 => {
  let {
    prefix
  } = _ref2;
  return {
    capsuleBase: {
      "props": {
        "sj55zd": _ueh72r.default
      },
      "alias": `${prefix}__tab-bar-more__capsule`
    },
    capsuleHovered: {
      "props": {
        "Jwef8y": _uf0hpal.default,
        "B2zomw9": _u8lm1xd.default,
        "Fgk69s": _u1fzbsm.default,
        "gu71dq": _u1x6icpj.default,
        "Btmb5he": _uoa5q.default
      }
    },
    capsulePressed: {
      "props": {
        "De3pzq": _u1xrgxpq.default,
        "B4j52fo": _u1cpcpb.default,
        "Bekrc4i": _u10qn3go.default,
        "Bn0qgzm": _uflkg.default,
        "ibv6hh": _uvn5txf.default,
        "icvyot": _us9j13g.default,
        "vrafjx": _umt7kug.default,
        "oivjwe": _uh3lwu.default,
        "wvpqe5": _ur.default,
        "g2u3we": _uwlx52q.default,
        "h3c5rm": _u10ismar.default,
        "B9xav0g": _u1g1nbrb.default,
        "zhjwy3": _uia0x1d.default,
        "Btl43ni": _u1773ck.default,
        "B7oj6ja": _upqf1qz.default,
        "Bbmb7ep": _u1cxaokx.default,
        "Beyfa6y": _uzx399i.default,
        "sj55zd": _uue.default
      },
      "alias": `${prefix}__tab-bar-more-capsule--active`
    }
  };
});
