// @universe-design/react/lib/tabs/Tab.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TabContent = exports.Tab = void 0;
var _u1n2z6e = _interopRequireDefault(require("./../.stylex/u1n2z6e7"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = require("@universe-design/stylex/react");
var _react2 = _interopRequireDefault(require("react"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _clsx = _interopRequireDefault(require("clsx"));
var _hooks = require("../overflow/hooks");
var _TabsContext = require("./TabsContext");
var _forwardRef = require("../util/forwardRef");
var _text = require("../text");
var _sx = require("./sx");
var _TabContext = require("./TabContext");
//.u11u7vat{min-width:20px;}
/* eslint-disable react/no-unknown-property */
/** @jsxRuntime classic */
/** @jsx jsx */
/** @jsxFrag React.Fragment */
const Tab = props => {
  const {
    size,
    panes,
    rootCssScope,
    type,
    tabBarGutter,
    handleTabClick,
    handleTabDoubleClick,
    activeTabIndex,
    activeKey,
    tabElementMapRef
  } = (0, _TabsContext.useTabsContext)();
  const tabStyles = (0, _sx.useTabStyles)({
    prefix: rootCssScope()
  });
  const {
    key
  } = (0, _TabContext.useTabContext)();
  const {
    omittedKeys
  } = (0, _hooks.useOverflowContext)();
  const paneProps = panes.find(p => p.key === key);
  const paneIndex = panes.findIndex(p => p.key === key);
  if (!paneProps) {
    return null;
  }
  const {
    tab,
    disabled
  } = paneProps;
  const isActive = paneIndex === activeTabIndex;
  const cls = (0, _clsx.default)(rootCssScope('__tab'), rootCssScope(`__tab--${type}`), {
    [rootCssScope('__tab--active')]: isActive,
    [rootCssScope('__tab--divider')]: type === 'card' && (paneIndex < activeTabIndex - 1 || paneIndex > activeTabIndex || paneIndex === activeTabIndex - 1 && omittedKeys.includes(activeKey)),
    [rootCssScope('__tab--disabled')]: disabled
  }, type === 'tag' && {
    [rootCssScope('-sm-border-radius')]: true,
    [rootCssScope('-neutral-background')]: !isActive,
    [rootCssScope('-primary-background')]: isActive,
    [rootCssScope('-color-pri-text')]: isActive
  });
  return (0, _react.jsx)("div", {
    ref: node => {
      var _tabElementMapRef$cur;
      (_tabElementMapRef$cur = tabElementMapRef.current) === null || _tabElementMapRef$cur === void 0 ? void 0 : _tabElementMapRef$cur.set(key, node);
    },
    style: {
      ...(type === 'line' ? {
        marginRight: tabBarGutter
      } : {})
    },
    className: cls,
    key: `tab-nav-key-${paneIndex}`,
    onClick: e => {
      var _props$onClick;
      (_props$onClick = props.onClick) === null || _props$onClick === void 0 ? void 0 : _props$onClick.call(props, e);
      handleTabClick(key, e);
    },
    onDoubleClick: e => handleTabDoubleClick(key, e),
    sx: [type !== 'capsule' && tabStyles.tabHover, (0, _sx.sxTabTypography)(size, isActive), (0, _sx.sxTabSize)(type, size), type === 'card' && [tabStyles.cardBase, isActive ? tabStyles.cardActive : tabStyles.cardDefault], type === 'capsule' && [tabStyles.capsuleBase, isActive ? tabStyles.capsuleActive : tabStyles.capsuleDefault], props.sx]
  }, tab);
};

// еӣ дёәеӨ–еұӮ TabItem жҳҜ flexпјҢжІЎжі•дҪҝз”Ё TextпјҢз”ЁиҝҷдёӘеҒҡеҚ•иЎҢзңҒз•Ҙ
exports.Tab = Tab;
const TabContent = exports.TabContent = (0, _forwardRef.forwardHTMLRef)((props, ref) => {
  const {
    isOmitted
  } = (0, _TabContext.useTabContext)();
  return (0, _react.jsx)(_text.Text, (0, _extends2.default)({
    as: "span",
    lineClamp: isOmitted ? undefined : 1,
    ref: ref
  }, (0, _react.useMergeComponentProps)(props, {
    sx: [{
      "props": {
        "Bf4jedk": _u1n2z6e.default
      }
    }]
  })));
});
