// @universe-design/react/lib/tabs/TabContext.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useTabContext = exports.useTab = exports.TabProvider = exports.TabContextProvider = void 0;
var React = _interopRequireWildcard(require("react"));
var _hooks = require("../overflow/hooks");
var _reactUtilities = require("@universe-design/react-utilities");
const useTab = props => {
  const {
    key
  } = props;
  const {
    omittedKeys
  } = (0, _hooks.useOverflowContext)();
  return React.useMemo(() => ({
    key,
    isOmitted: omittedKeys.includes(key)
  }), [key, omittedKeys]);
};
exports.useTab = useTab;
const {
  useContext: useTabContext,
  Provider: TabContextProvider
} = (0, _reactUtilities.createContextHook)('TabContext', {
  strict: true
});
exports.TabContextProvider = TabContextProvider;
exports.useTabContext = useTabContext;
const TabProvider = _ref => {
  let {
    tabKey,
    children
  } = _ref;
  return /*#__PURE__*/React.createElement(TabContextProvider, {
    value: useTab({
      key: tabKey
    })
  }, children);
};
exports.TabProvider = TabProvider;
