// @universe-design/react/lib/tabs/TabInk.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TabsInkBar = void 0;
var _clsx = _interopRequireDefault(require("clsx"));
var _react = _interopRequireWildcard(require("react"));
var _hooks = require("../hooks");
var _hooks2 = require("../overflow/hooks");
var _TabsContext = require("./TabsContext");
function getOffsetRect(el) {
  const {
    offsetLeft = 0,
    offsetTop = 0,
    offsetHeight = 0,
    offsetWidth = 0
  } = el !== null && el !== void 0 ? el : {};
  return {
    left: offsetLeft,
    top: offsetTop,
    height: offsetHeight,
    width: offsetWidth,
    bottom: offsetTop + offsetHeight,
    right: offsetLeft + offsetWidth
  };
}

/**
 * Tab 标签下划线
 */
const TabsInkBar = _ref => {
  let {
    offsetY
  } = _ref;
  const {
    type,
    rootCssScope,
    activeKey,
    panes,
    tabElementMapRef,
    tabBarWrapperRef,
    moreTabRef
  } = (0, _TabsContext.useTabsContext)();
  const {
    omittedKeys,
    itemWidths,
    restWidth,
    containerWidth
  } = (0, _hooks2.useOverflowContext)();
  const inkRef = (0, _react.useRef)(null);
  const getInkStyle = (0, _react.useCallback)(() => {
    // overflow item 每个元素的变化都会触发这个元素 rerender
    const tabBarWrapper = tabBarWrapperRef.current;
    if (!tabBarWrapper) {
      return {};
    }
    const inkOffsetRect = getOffsetRect(inkRef.current);
    for (const pane of panes) {
      if (omittedKeys.includes(pane.key)) {
        const moreTab = moreTabRef.current;
        const moreOffsetRect = getOffsetRect(moreTab);
        return {
          width: moreOffsetRect.width,
          left: moreOffsetRect.left,
          // 多减去 1px 是为了（蓝色）选中线在（灰色）分割线上面紧贴合
          top: moreOffsetRect.bottom - inkOffsetRect.height + offsetY
        };
      }
      if (pane.key === activeKey) {
        var _tabElementMapRef$cur;
        const tab = (_tabElementMapRef$cur = tabElementMapRef.current) === null || _tabElementMapRef$cur === void 0 ? void 0 : _tabElementMapRef$cur.get(activeKey);
        const tabOffsetRect = getOffsetRect(tab);
        return {
          width: tabOffsetRect.width,
          left: tabOffsetRect.left,
          // 多减去 1px 是为了（蓝色）选中线在（灰色）分割线上面紧贴合
          top: tabOffsetRect.bottom - inkOffsetRect.height + offsetY
        };
      }
    }
    return {};
  }, [activeKey, moreTabRef, offsetY, omittedKeys, panes, tabBarWrapperRef, tabElementMapRef]);
  const [inkStyle, setInkStyle] = (0, _react.useState)(getInkStyle);

  /**
   * 保证在处于稳定态的时候才加渐变动画，这样默认的 activeKey 不再最左侧时，首次加载不会有一个飞过去的效果
   */
  const [animated, setAnimated] = (0, _react.useState)(false);
  const debouncedSetAnimated = (0, _hooks.useDebouncedCallback)((0, _react.useCallback)(() => {
    setAnimated(true);
  }, []),
  // eslint-disable-next-line @typescript-eslint/no-magic-numbers
  800);
  _react.default.useEffect(() => {
    // 为了减少一些不必要的 resize observer 所以这里直接通过 overflow item 宽度变换来代替真实元素的尺寸变化

    const newStyle = getInkStyle();
    /**
     * 如果使用 display: none 控制元素显影，不重置 style
     * 防止其位置抖动
     **/
    if (newStyle.width === 0) {
      return;
    }
    setInkStyle(newStyle);
    debouncedSetAnimated();
  }, [containerWidth, debouncedSetAnimated, getInkStyle, itemWidths, restWidth, tabBarWrapperRef]);
  if (type === 'line') {
    const inkCls = (0, _clsx.default)(rootCssScope('__tab-bar-ink'), animated && rootCssScope('__tab-bar-ink-animated'));
    return /*#__PURE__*/_react.default.createElement("div", {
      style: inkStyle,
      className: inkCls,
      ref: inkRef
    });
  }
  return null;
};
exports.TabsInkBar = TabsInkBar;
