// @universe-design/react/lib/tabs/Tabs.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Tabs = exports.TabMoreContent = exports.TabMoreArrow = exports.TabMore = void 0;
var _u98ahtk = _interopRequireDefault(require("./../.stylex/u98ahtk"));
var _u1suju4n = _interopRequireDefault(require("./../.stylex/u1suju4n"));
var _u19ubm5t = _interopRequireDefault(require("./../.stylex/u19ubm5t"));
var _umy11dx = _interopRequireDefault(require("./../.stylex/umy11dx"));
var _u19b90lh = _interopRequireDefault(require("./../.stylex/u19b90lh"));
var _u1vk5ola = _interopRequireDefault(require("./../.stylex/u1vk5ola"));
var _umu3nk = _interopRequireDefault(require("./../.stylex/umu3nk9"));
var _u1981n0b = _interopRequireDefault(require("./../.stylex/u1981n0b"));
var _u1to0gca = _interopRequireDefault(require("./../.stylex/u1to0gca"));
var _u5jxv = _interopRequireDefault(require("./../.stylex/u5jxv71"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _react = require("@universe-design/stylex/react");
var _DownBoldOutlined = _interopRequireDefault(require("@universe-design/icons-react/DownBoldOutlined"));
var _stylex2 = require("../stylex");
var _clsx = _interopRequireDefault(require("clsx"));
var _react2 = _interopRequireWildcard(require("react"));
var _hooks = require("../config-provider/hooks");
var _dropdown = require("../dropdown");
var _menu = require("../menu");
var _overflow = require("../overflow");
var _hooks2 = require("../overflow/hooks");
var _util = require("../util");
var _TabPane = require("./TabPane");
var _TabsContext = require("./TabsContext");
var _sx = require("./sx");
var _Tab = require("./Tab");
var _TabContext = require("./TabContext");
var _ScrollableTabsHeader = require("./ScrollableTabsHeader");
var _TabInk = require("./TabInk");
require("./style");
var _sizing_behavior = require("./sizing_behavior");
//.u1acs6jw{width:fit-content;}
//.u9w1357{background:@N900-5;}
//.u96dupq{border-top-left-radius:@border-radius-sm;}
//.u1msb61c{border-top-right-radius:@border-radius-sm;}
//.u3rkx2v{border-bottom-right-radius:@border-radius-sm;}
//.u17hi4zz{border-bottom-left-radius:@border-radius-sm;}
//.u1xg1ack{position:sticky;}
//.u113wtx2{top:0px;}
//.u11fbdcc{background-color:@bg-body;}
//.u10pi13n{position:relative;}
/* eslint-disable react/no-unknown-property */
/** @jsxRuntime classic */
/** @jsx jsx */
/** @jsxFrag React.Fragment */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const TabMoreArrow = () => (0, _react.jsx)(_DownBoldOutlined.default, null);
exports.TabMoreArrow = TabMoreArrow;
const TabMoreContent = () => {
  const locale = (0, _hooks.useLocaleReceiver)('tabs');
  return (0, _react.jsx)("span", null, locale.more);
};

/**
 * `<TabMore />` 组件的属性
 * @public
 */
exports.TabMoreContent = TabMoreContent;
const TabMore = _ref => {
  var _ref3;
  let {
    children,
    ...dropDownProps
  } = _ref;
  const {
    panes,
    rootCssScope,
    handleTabClick,
    activeKey,
    moreTabRef,
    size,
    type
  } = (0, _TabsContext.useTabsContext)();
  const {
    omittedKeys
  } = (0, _hooks2.useOverflowContext)();
  const [showMoreTabs, setShowMoreTabs] = (0, _react2.useState)(false);
  const isSelected = omittedKeys.includes(activeKey);
  const mergedVisible = _util.__SCREENSHOT__ ? isSelected : showMoreTabs && Boolean(omittedKeys.length);
  const tabsMoreStyles = (0, _sx.useTabsMoreStyles)({
    prefix: rootCssScope()
  });
  const morePanes = panes.filter(p => omittedKeys.includes(p.key));
  const cls = (0, _clsx.default)(rootCssScope('__tab-bar-more'), rootCssScope(`__tab-bar-more--${type}`), {
    [rootCssScope('__tab-bar-more--selected')]: isSelected,
    [rootCssScope('__tab-bar-more--show')]: mergedVisible
  });
  const menu = (0, _react.jsx)(_menu.Menu, {
    selectedKeys: [activeKey],
    textOverflow: false,
    selectable: true
  }, morePanes.map(pane => {
    const {
      disabled,
      key,
      tab
    } = pane;
    return (0, _react.jsx)(_menu.Menu.Item, {
      key: key,
      disabled: disabled,
      onClick: _ref2 => {
        let {
          domEvent
        } = _ref2;
        return handleTabClick(key, domEvent);
      }
    }, (0, _react.jsx)(_TabContext.TabProvider, {
      tabKey: key
    }, tab));
  }));
  const moreTabContentNode = (0, _react.jsx)(_react2.default.Fragment, null, (0, _react.jsx)(TabMoreContent, null), (0, _react.jsx)(TabMoreArrow, null));
  return (0, _react.jsx)("div", {
    className: (0, _clsx.default)(rootCssScope('__tab-bar__operation'))
  }, (0, _react.jsx)("div", {
    className: cls,
    sx: [!isSelected ? _stylex2.cssTypographyTitle5 : _stylex2.cssTypographyTitle4]
  }, (0, _react.jsx)(_dropdown.Dropdown, (0, _extends2.default)({
    trigger: ['hover', 'click']
  }, dropDownProps, {
    overlay: menu,
    visible: mergedVisible,
    onVisibleChange: setShowMoreTabs
  }), (0, _react.jsx)("div", {
    ref: moreTabRef,
    className: (0, _clsx.default)(rootCssScope('__tab-bar-more-body'), type === 'tag' && [rootCssScope('-sm-border-radius'), isSelected ? [rootCssScope('-primary-background'), rootCssScope('-color-pri-text')] : rootCssScope('-neutral-background')]),
    sx: [(0, _sx.sxTabTypography)(size, isSelected), (0, _sx.sxTabSize)(type, size), type === 'capsule' && [tabsMoreStyles.capsuleBase, !isSelected && tabsMoreStyles.capsuleHovered, isSelected && tabsMoreStyles.capsulePressed]]
  }, (_ref3 = (0, _util.isFunction)(children) ? children({
    omittedKeys: omittedKeys
  }) : children) !== null && _ref3 !== void 0 ? _ref3 : moreTabContentNode))));
};
exports.TabMore = TabMore;
const Tabs = exports.Tabs = /*#__PURE__*/(0, _react2.forwardRef)((props, ref) => {
  const {
    tabMore,
    sticky
  } = props;
  const tabCtx = (0, _TabsContext.useTabs)(props);
  const {
    style,
    className,
    tabBarStyle,
    type = 'line',
    size = 'medium',
    maxCount,
    tabBarExtraContent,
    rootCssScope,
    tabBarWrapperRef,
    panes,
    rootTabsSizingBehavior,
    scrollable
  } = tabCtx;
  const usedSize = (0, _util.getDefaultSizeAbbr)(size, 'medium');
  const tabCls = (0, _clsx.default)(rootCssScope(), {
    [rootCssScope(`--type-${type}`)]: type,
    [rootCssScope(`--size-${usedSize}`)]: usedSize
  }, className);
  const tabsRef = (0, _react2.useRef)(null);
  // 蓝色选中线和 divider 重合
  const inkOffsetY = 0;
  (0, _react2.useImperativeHandle)(ref, () => tabsRef.current);
  const tabsInkBarNode = (0, _react2.useMemo)(() => (0, _react.jsx)(_TabInk.TabsInkBar, {
    offsetY: inkOffsetY
  }), [inkOffsetY]);
  const moreTabNode = (0, _react2.useMemo)(() => tabMore !== null && tabMore !== void 0 ? tabMore : (0, _react.jsx)(TabMore, null), [tabMore]);
  const renderLeftExtraContentNode = (0, _react2.useMemo)(() => {
    const assertExtra = tabBarExtraContent;
    if (!assertExtra) {
      return null;
    }
    return (0, _react.jsx)("div", {
      className: rootCssScope('__tab-extra-content')
    }, assertExtra.left);
  }, [rootCssScope, tabBarExtraContent]);
  const renderRightExtraContentNode = (0, _react2.useMemo)(() => {
    const assertExtra = tabBarExtraContent;
    if (!assertExtra) {
      return null;
    }
    return (0, _react.jsx)("div", {
      className: rootCssScope('__tab-extra-content')
    }, assertExtra.right || !assertExtra.left && assertExtra || null);
  }, [rootCssScope, tabBarExtraContent]);
  const getItemStyle = _react2.default.useCallback(() => (0, _sizing_behavior.resolveSizingBehavior)(rootTabsSizingBehavior), [rootTabsSizingBehavior]);
  return (0, _react.jsx)(_TabsContext.TabsContext.Provider, {
    value: tabCtx
  }, (0, _react.jsx)("div", {
    ref: tabsRef,
    style: style,
    className: tabCls
  }, (() => (0, _react.jsx)("div", {
    className: (0, _clsx.default)(rootCssScope('__tab-bar-holder'), rootCssScope('__tab-bar-nav'), sticky && rootCssScope('__tab-bar-nav--sticky')),
    sx: [{
      "props": {
        "qhf8xq": _u5jxv.default
      }
    }, sticky && {
      "props": {
        "qhf8xq": _umu3nk.default,
        "Bhzewxz": _u1981n0b.default,
        "De3pzq": _u1to0gca.default
      }
    }, type === 'capsule' && {
      "props": {
        "ayd6f0": _u1suju4n.default,
        "Btl43ni": _u19ubm5t.default,
        "B7oj6ja": _umy11dx.default,
        "Bbmb7ep": _u19b90lh.default,
        "Beyfa6y": _u1vk5ola.default
      }
    },
    // scrollable 时会撑开父级，无法滚动 && responsive 时无法正常展开，只有一个 more
    type === 'capsule' && !scrollable && maxCount !== 'responsive' && {
      "props": {
        "a9b677": _u98ahtk.default
      }
    }]
  }, renderLeftExtraContentNode, scrollable ? (0, _react.jsx)(_ScrollableTabsHeader.ScrollableTabsHeader, {
    ref: tabBarWrapperRef,
    inkOffsetY: inkOffsetY
  }) : (0, _react.jsx)("div", {
    ref: tabBarWrapperRef,
    className: rootCssScope('__tab-bar-wrap')
  }, (0, _react.jsx)(_overflow.Overflow, {
    style: tabBarStyle,
    getItemStyle: getItemStyle,
    restStyle: (0, _sizing_behavior.resolveSizingBehavior)(rootTabsSizingBehavior),
    keys: panes.map(p => p.key),
    maxCount: maxCount,
    itemRenderer: key => (0, _react.jsx)(_TabContext.TabProvider, {
      tabKey: key
    }, (0, _react.jsx)(_Tab.Tab, null)),
    restRenderer: () => moreTabNode,
    addonRenderer: () => tabsInkBarNode,
    alwaysObserveResize: true
  })), renderRightExtraContentNode, type === 'line' && (0, _react.jsx)("div", {
    className: rootCssScope('__tab-bar-holder-divider')
  })))(), (() => (0, _react.jsx)("div", {
    className: rootCssScope('__content')
  }, panes.map(paneProps => (0, _react.jsx)(_TabPane.TabPane, (0, _extends2.default)({}, paneProps, {
    internalKey: paneProps.key,
    key: paneProps.key
  })))))()));
});
if (process.env.NODE_ENV !== "production") {
  Tabs.displayName = 'Tabs';
}
