// @universe-design/react/lib/tabs/TabsContext.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UD_TABPANE_ID = exports.TabsContext = void 0;
exports.isTabPane = isTabPane;
exports.useTabsContext = exports.useTabs = void 0;
var React = _interopRequireWildcard(require("react"));
var _useContextSelector = require("use-context-selector");
var _reactUtilities = require("@universe-design/react-utilities");
var _hooks = require("../config-provider/hooks");
var _util = require("../util");
/**
 * Tab 尺寸行为类型
 * 定义单个 Tab 的尺寸控制方式
 * 支持简单字符串类型和详细的 SizingBehavior 对象类型
 */

const UD_TABPANE_ID = exports.UD_TABPANE_ID = 'UD_TABPANE_ID';
function isTabPane(node) {
  var _node$type;
  return node === null || node === void 0 ? void 0 : (_node$type = node.type) === null || _node$type === void 0 ? void 0 : _node$type[UD_TABPANE_ID];
}

/**
 * @public
 */

// 'editable-card' 类型暂缓提供

/**
 * `<Tabs />` 组件的属性
 * @public
 */

/**
 * `<Tabs.Pane>` 组件的属性
 * @public
 */

const TabsContext = exports.TabsContext = (0, _useContextSelector.createContext)({});
if (process.env.NODE_ENV !== "production") {
  TabsContext.displayName = 'TabsContext';
}
const useTabs = props => {
  const {
    style,
    className,
    children,
    type = 'line',
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    tabBarGutter = type === 'tag' ? 16 : 28,
    scrollButtonGutter = type === 'line' ? 28 : 0,
    tabBarStyle,
    activeKey: actKeyInProps,
    defaultActiveKey,
    size = 'medium',
    maxCount: maxCountProp,
    tabBarExtraContent,
    onChange,
    onTabClick,
    onTabDoubleClick,
    scrollable: scrollableProp,
    tabsSizingBehavior
  } = props;
  const maxCount = (() => {
    if (!maxCountProp && type === 'capsule') {
      return Infinity;
    }
    if (!maxCountProp) {
      return 'responsive';
    }
    return maxCountProp;
  })();
  const scrollable = (() => // if (type === 'capsule') {
  //   warning(!scrollableProp, devWarning('Tabs', 'scrollable is not supported in capsule type'));
  //   return false;
  // }
  scrollableProp !== null && scrollableProp !== void 0 ? scrollableProp : false)();
  const rootTabsSizingBehavior = React.useMemo(() => {
    if (!tabsSizingBehavior || tabsSizingBehavior === 'hug') {
      return {
        mode: 'hug',
        maxWidth: 240
      };
    }
    if (tabsSizingBehavior === 'flex') {
      return {
        mode: 'flex',
        maxWidth: 240
      };
    }
    return tabsSizingBehavior;
  }, [tabsSizingBehavior]);
  const panes = React.useMemo(() => (0, _util.toFlatArray)(children).filter(isTabPane).map((child, index) => ({
    ...child.props,
    key: child.key && String(child.key) || `${index}`
  })), [children]);
  const [activeKey, setActiveKey] = (0, _reactUtilities.useControlledState)(defaultActiveKey || panes[0].key, actKeyInProps);
  const handleTabClick = React.useCallback((key, e) => {
    const pane = panes.find(p => p.key === key);
    if (!pane || pane.disabled) {
      return;
    }
    setActiveKey(key);
    (0, _util.isFunction)(onChange) && onChange(key);
    (0, _util.isFunction)(onTabClick) && onTabClick(key, e);
  }, [onChange, onTabClick, panes, setActiveKey]);
  const handleTabDoubleClick = React.useCallback((key, e) => {
    const pane = panes.find(p => p.key === key);
    if (!pane || pane.disabled) {
      return;
    }
    if ((0, _util.isFunction)(onTabDoubleClick)) {
      onTabDoubleClick(key, e);
    }
  }, [onTabDoubleClick, panes]);
  const rootCssScope = (0, _hooks.useCssScope)('tabs');
  const activeTabIndex = panes.findIndex(p => p.key === activeKey);
  const tabElementMapRef = React.useRef(new Map());
  // position relative 的元素，用于定位 ink
  const tabBarWrapperRef = React.useRef(null);
  // 更多 tab
  const moreTabRef = React.useRef(null);
  return React.useMemo(() => ({
    style,
    className,
    children,
    tabBarGutter,
    tabBarStyle: {
      // avoid browser's text selection behavior, when double click
      userSelect: (0, _util.isFunction)(onTabDoubleClick) ? 'none' : 'auto',
      ...tabBarStyle,
      ...(scrollable ? {
        flexWrap: 'nowrap'
      } : {})
    },
    activeKey,
    type,
    size,
    maxCount,
    tabBarExtraContent,
    onChange,
    onTabClick,
    onTabDoubleClick,
    rootCssScope,
    setActiveKey,
    handleTabClick,
    handleTabDoubleClick,
    panes,
    activeTabIndex,
    tabElementMapRef,
    tabBarWrapperRef,
    moreTabRef,
    scrollButtonGutter,
    rootTabsSizingBehavior,
    scrollable
  }), [activeKey, activeTabIndex, children, className, handleTabClick, handleTabDoubleClick, maxCount, onChange, onTabClick, onTabDoubleClick, panes, rootCssScope, scrollButtonGutter, scrollable, setActiveKey, size, style, tabBarExtraContent, tabBarGutter, tabBarStyle, type, rootTabsSizingBehavior]);
};
exports.useTabs = useTabs;
const useTabsContext = () => (0, _useContextSelector.useContext)(TabsContext);
exports.useTabsContext = useTabsContext;
