// @universe-design/react/es/image-viewer/components/Download.js
import * as React from 'react';
import clsx from 'clsx';
import DownloadOutlined from '@universe-design/icons-react/DownloadOutlined';
import { useLocaleReceiver } from '../../config-provider/hooks';
import { useImageViewerContext } from '../ImageViewerContext';
import Tooltip from '../../tooltip';
import { IconButton } from '../../button';
import '../style';
function downloadByLink(blobUrl, name) {
  const a = document.createElement('a');
  a.style.display = 'none';
  a.href = blobUrl;
  a.download = name;
  document.body.appendChild(a);
  a.click();
  document.body.removeChild(a);
}
const downloadBlob = async (blob, name) => {
  const blobUrl = window.URL.createObjectURL(blob);
  downloadByLink(blobUrl, name);
  window.URL.revokeObjectURL(blobUrl);
};
const downloadImage = async (source, name) => {
  if (source instanceof Blob) {
    await downloadBlob(source, name);
    return;
  }
  if (source.startsWith('blob:')) {
    downloadByLink(source, name);
    return;
  }
  const blob = await fetch(source).then(r => r.blob());
  await downloadBlob(blob, name);
};

/**
 * 图片查看器下载工具 hook
 *
 * @public
 */
export const useImageViewerToolbarDownload = () => {
  const {
    images,
    currentIndex
  } = useImageViewerContext();
  const download = React.useCallback(async () => {
    var _images$currentIndex$;
    await downloadImage(images[currentIndex].src, (_images$currentIndex$ = images[currentIndex].alt) !== null && _images$currentIndex$ !== void 0 ? _images$currentIndex$ : `${new Date().getTime()}.jpg`);
  }, [currentIndex, images]);
  return {
    download
  };
};
export const Download = /*#__PURE__*/React.memo(_ref => {
  let {
    onDownload
  } = _ref;
  const locale = useLocaleReceiver('imageViewer');
  const {
    rootCssScope
  } = useImageViewerContext();
  const {
    images,
    currentIndex
  } = useImageViewerContext();
  const {
    download
  } = useImageViewerToolbarDownload();
  return /*#__PURE__*/React.createElement(Tooltip, {
    title: locale.download,
    overlayClassName: rootCssScope('__toolbar__tooltip')
  }, /*#__PURE__*/React.createElement(IconButton, {
    size: 20,
    icon: /*#__PURE__*/React.createElement(DownloadOutlined, null),
    className: clsx(rootCssScope('__toolbar__button'), rootCssScope('__toolbar__download')),
    onClick: onDownload ? () => onDownload(images[currentIndex], {
      downloadImage
    }) : download
  }));
});
