// @universe-design/react/es/image-viewer/components/Pagination.js
//.u22iagw{display:flex;}
import ukchsq7 from "./../../.stylex/ukchsq7";
//.u4d9j23{justify-content:center;}
import ut7q8es from "./../../.stylex/ut7q8es";
//.uz5stix{white-space:nowrap;}
import uugkpzq from "./../../.stylex/uugkpzq";
/* eslint-disable react/no-unknown-property */
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx, makeStylexStyles } from '@universe-design/stylex/react';
import stylex from '@universe-design/stylex';
import * as React from 'react';
import LeftOutlined from '@universe-design/icons-react/LeftOutlined';
import RightOutlined from '@universe-design/icons-react/RightOutlined';
import clsx from 'clsx';
import { useEventCallback, KeyboardKey } from '@universe-design/react-utilities';
import { useLocaleReceiver } from '../../config-provider/hooks';
import { useImageViewerContext } from '../ImageViewerContext';
import Tooltip from '../../tooltip';
import { IconButton } from '../../button';
import { Space } from '../../space';
import '../style';
const usePaginationStylexStyles = makeStylexStyles(_ref => {
  let {
    prefix
  } = _ref;
  return {
    info: {
      "props": {
        "mc9l5x": ukchsq7,
        "Brf1p80": ut7q8es,
        "Huce71": uugkpzq
      },
      "alias": `${prefix}__info`
    }
  };
});

/**
 * 图片查看器翻页工具 hook
 *
 * @public
 */
export const useImageViewerToolbarPagination = () => {
  const {
    images,
    setCurrentIndex,
    currentIndex
  } = useImageViewerContext();
  const paginateForwardDisabled = currentIndex === images.length - 1;
  const paginateBackward = useEventCallback(() => {
    if (currentIndex > 0) {
      setCurrentIndex(currentIndex - 1);
    }
  });
  const paginateForward = useEventCallback(() => {
    if (currentIndex < images.length - 1) {
      setCurrentIndex(currentIndex + 1);
    }
  });
  React.useEffect(() => {
    const onKeydown = event => {
      if (event.key === KeyboardKey.ArrowLeft) {
        paginateBackward();
      } else if (event.key === KeyboardKey.ArrowRight) {
        paginateForward();
      }
    };
    document.addEventListener('keydown', onKeydown);
    return () => {
      document.removeEventListener('keydown', onKeydown);
    };
  }, [paginateBackward, paginateForward]);
  return {
    paginateBackwardDisabled: currentIndex === 0,
    paginateForwardDisabled,
    paginateBackward,
    paginateForward
  };
};
export const Pagination = () => {
  const locale = useLocaleReceiver('imageViewer');
  const {
    images,
    currentIndex,
    rootCssScope
  } = useImageViewerContext();
  const styles = usePaginationStylexStyles({
    prefix: rootCssScope('__pagination')
  });
  const {
    paginateBackward,
    paginateBackwardDisabled,
    paginateForward,
    paginateForwardDisabled
  } = useImageViewerToolbarPagination();
  return jsx(Space, {
    size: 12
  }, jsx(Tooltip, {
    title: locale.prevArrow,
    overlayClassName: rootCssScope('__toolbar__tooltip'),
    wrapDisabledChildren: true
  }, jsx(IconButton, {
    size: 20,
    disabled: paginateBackwardDisabled,
    icon: jsx(LeftOutlined, null),
    className: clsx(rootCssScope('__toolbar__button'), paginateBackwardDisabled && rootCssScope('__toolbar__button-disabled')),
    onClick: paginateBackward
  })), jsx("div", {
    sx: [styles.info]
  }, `${currentIndex + 1} / ${images.length}`), jsx(Tooltip, {
    title: locale.nextArrow,
    wrapDisabledChildren: true,
    overlayClassName: rootCssScope('__toolbar__tooltip')
  }, jsx(IconButton, {
    disabled: paginateForwardDisabled,
    size: 20,
    icon: jsx(RightOutlined, null),
    className: clsx(rootCssScope('__toolbar__button'), paginateForwardDisabled && rootCssScope('__toolbar__button-disabled')),
    onClick: paginateForward
  })));
};
