// @universe-design/react/es/image-viewer/components/Scaling.js
//.ueqmc2u{width:40px;}
import uwf09ar from "./../../.stylex/uwf09ar";
//.u22iagw{display:flex;}
import ukchsq7 from "./../../.stylex/ukchsq7";
//.u4d9j23{justify-content:center;}
import ut7q8es from "./../../.stylex/ut7q8es";
/* eslint-disable @typescript-eslint/no-magic-numbers */
/* eslint-disable react/no-unknown-property */
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx, makeStylexStyles } from '@universe-design/stylex/react';
import stylex from '@universe-design/stylex';
import * as React from 'react';
import OriginalSizeOutlined from '@universe-design/icons-react/OriginalSizeOutlined';
import ResizeOutlined from '@universe-design/icons-react/ResizeOutlined';
import ZoomInOutlined from '@universe-design/icons-react/ZoomInOutlined';
import ZoomOutOutlined from '@universe-design/icons-react/ZoomOutOutlined';
import { AMPLIFY_NUMBER, SCALING_SPEED } from '../constants';
import clsx from 'clsx';
import { useLocaleReceiver } from '../../config-provider/hooks/index';
import { KeyboardKey } from '@universe-design/react-utilities';
import { IconButton } from '../../button/index';
import { Tooltip } from '../../tooltip/index';
import { useEventCallback } from '../../hooks/index';
import { useImageViewerContext } from '../ImageViewerContext';
import { Space } from '../../space';
import '../style/index';
const useScalingStylexStyles = makeStylexStyles(_ref => {
  let {
    prefix
  } = _ref;
  return {
    info: {
      "props": {
        "a9b677": uwf09ar,
        "mc9l5x": ukchsq7,
        "Brf1p80": ut7q8es
      },
      "alias": `${prefix}__info`
    }
  };
});

/**
 * 图片查看器缩放工具 hook
 *
 * @public
 */
export const useImageViewerToolbarScaling = () => {
  // 从上下文获取图片查看器的缩放相关状态和方法
  const {
    scale,
    initTransform,
    scaleByOrigin,
    minScale,
    maxScale
  } = useImageViewerContext();

  // 适应屏幕状态，控制是否自动适配容器
  const [fitToScreen, setFitToScreen] = React.useState(true);

  // 切换适应屏幕状态的回调函数
  const toggleFitToScreen = useEventCallback(() => {
    setFitToScreen(!fitToScreen);
    initTransform({
      fitToContainer: !fitToScreen
    });
  });

  // 计算缩放按钮的禁用状态（保留两位小数避免浮点数精度问题）
  const scaleDownDisabled = Number(scale.toFixed(2)) <= minScale;
  const scaleUpDisabled = Number(scale.toFixed(2)) >= maxScale;

  // 放大操作处理函数
  const scaleUp = useEventCallback(() => {
    scaleByOrigin({
      getNextScale: currentScale => currentScale * (1 + SCALING_SPEED)
    });
  });

  // 缩小操作处理函数
  const scaleDown = useEventCallback(() => {
    scaleByOrigin({
      getNextScale: currentScale => currentScale * (1 - SCALING_SPEED)
    });
  });

  /**
   * Scaling
   * - 键盘操作（上下键）
   */
  React.useEffect(() => {
    /**
     * 上、下键缩放，基于容器中心缩放
     */
    const handleKeyDown = event => {
      if (event.key === KeyboardKey.ArrowUp) {
        scaleUp();
      } else if (event.key === KeyboardKey.ArrowDown) {
        scaleDown();
      }
    };
    document.addEventListener('keydown', handleKeyDown);
    return () => {
      document.removeEventListener('keydown', handleKeyDown);
    };
  }, [scaleDown, scaleUp]);
  return {
    fitToScreen,
    toggleFitToScreen,
    scaleDownDisabled,
    scaleUpDisabled,
    scaleUp,
    scaleDown
  };
};
export const Scaling = () => {
  const locale = useLocaleReceiver('imageViewer');
  const {
    scale,
    rootCssScope
  } = useImageViewerContext();
  const styles = useScalingStylexStyles({
    prefix: rootCssScope('__scaling')
  });
  const {
    fitToScreen,
    toggleFitToScreen,
    scaleDownDisabled,
    scaleUpDisabled,
    scaleUp,
    scaleDown
  } = useImageViewerToolbarScaling();
  return jsx(Space, {
    size: 12,
    className: rootCssScope('__scaling')
  }, jsx(Tooltip, {
    title: locale.zoomOut,
    wrapDisabledChildren: true,
    overlayClassName: rootCssScope('__toolbar__tooltip')
  }, jsx(IconButton, {
    size: 20,
    disabled: scaleDownDisabled,
    icon: jsx(ZoomOutOutlined, null),
    className: clsx(rootCssScope('__toolbar__button'), scaleDownDisabled && rootCssScope('__toolbar__button-disabled')),
    onClick: scaleDown
  })), jsx("div", {
    sx: [styles.info]
  }, `${(scale * AMPLIFY_NUMBER).toFixed(0)}%`), jsx(Tooltip, {
    title: locale.zoomIn,
    wrapDisabledChildren: true,
    overlayClassName: rootCssScope('__toolbar__tooltip')
  }, jsx(IconButton, {
    disabled: scaleUpDisabled,
    size: 20,
    icon: jsx(ZoomInOutlined, null),
    className: clsx(rootCssScope('__toolbar__button'), scaleUpDisabled && rootCssScope('__toolbar__button-disabled')),
    onClick: scaleUp
  })), jsx(Tooltip, {
    title: fitToScreen ? locale.originSize : locale.bestFit,
    overlayClassName: rootCssScope('__toolbar__tooltip')
  }, jsx(IconButton, {
    size: 20,
    icon: fitToScreen ? jsx(OriginalSizeOutlined, null) : jsx(ResizeOutlined, null),
    className: rootCssScope('__toolbar__button'),
    onClick: toggleFitToScreen
  })));
};
