// @universe-design/styles/cjs/ccm-ccm-sheet/index.dark.atom-raw.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ccmtoken_sheet_minichart_professional_red_raw = exports.ccmtoken_sheet_minichart_professional_blue_raw = exports.ccmtoken_sheet_minichart_pastel_yellow_raw = exports.ccmtoken_sheet_minichart_pastel_blue_raw = exports.ccmtoken_sheet_minichart_fresh_yellow_raw = exports.ccmtoken_sheet_minichart_fresh_cyan_raw = exports.ccmtoken_sheet_minichart_bright_yellow_raw = exports.ccmtoken_sheet_minichart_bright_purple_raw = void 0;
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
exports.ccmtoken_sheet_minichart_bright_purple_raw = '98, 75, 239';
exports.ccmtoken_sheet_minichart_bright_yellow_raw = '237, 177, 24';
exports.ccmtoken_sheet_minichart_fresh_cyan_raw = '31, 208, 218';
exports.ccmtoken_sheet_minichart_fresh_yellow_raw = '238, 191, 38';
exports.ccmtoken_sheet_minichart_pastel_blue_raw = '76, 146, 173';
exports.ccmtoken_sheet_minichart_pastel_yellow_raw = '190, 164, 26';
exports.ccmtoken_sheet_minichart_professional_blue_raw = '43, 75, 138';
exports.ccmtoken_sheet_minichart_professional_red_raw = '151, 46, 43';
