// @universe-design/react/lib/shared/components/expand-button/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ExpandButton = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _ExpandDownFilled = require("./ExpandDownFilled");
var _DownBoldOutlined = require("./DownBoldOutlined");
var _clsx = _interopRequireDefault(require("clsx"));
var _button = require("../../../button");
var _hooks = require("../../../config-provider/hooks");
var _util = require("../../../util");
require("./index.less");
const ExpandButton = exports.ExpandButton = /*#__PURE__*/_react.default.forwardRef((_ref, ref) => {
  let {
    isExpanded,
    icon,
    iconType,
    size = iconType === 'ExpandDownFilled' ? 'xxs' : 'small',
    onExpandChange,
    onClick,
    className,
    asPlaceholder = false,
    readonly = false,
    ...restProps
  } = _ref;
  const cssScope = (0, _hooks.useCssScope)('expandButton');
  const handleClick = _react.default.useCallback(e => {
    (0, _util.isFunction)(onClick) && !readonly && onClick(e);
    (0, _util.isFunction)(onExpandChange) && onExpandChange(!isExpanded);
  }, [isExpanded, onClick, onExpandChange, readonly]);
  const iconNode = (() => {
    if ((0, _util.isFunction)(icon)) {
      return icon({
        isExpanded
      });
    }
    if (typeof icon !== 'undefined') {
      return icon;
    }
    if (iconType === 'ExpandDownFilled') {
      return /*#__PURE__*/_react.default.createElement(_ExpandDownFilled.ExpandableExpandDownFilled, {
        isExpanded: isExpanded
      });
    }
    if (iconType === 'DownBoldOutlined') {
      return /*#__PURE__*/_react.default.createElement(_DownBoldOutlined.ExpandableDownBoldOutlined, {
        isExpanded: isExpanded
      });
    }
  })();
  return /*#__PURE__*/_react.default.createElement(_button.Button, (0, _extends2.default)({
    type: "icon",
    ref: ref,
    className: (0, _clsx.default)(cssScope(), className, asPlaceholder && cssScope('-as-placeholder'), readonly && cssScope('-readonly'), size === 'xxs' && cssScope('-xxs')),
    icon: iconNode,
    onClick: handleClick,
    size: size === 'xxs' ? undefined : size
  }, restProps));
});
