// @universe-design/react-utilities/lib-commonjs/util/assert.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getWindow = getWindow;
exports.isFunction = isFunction;
exports.isNonNullable = isNonNullable;
exports.isPromise = exports.isObject = exports.isNonReplacedElement = void 0;
exports.isReactHTMLElement = isReactHTMLElement;
exports.isReactInputElement = isReactInputElement;
exports.isString = exports.isSVG = exports.isReplacedElement = void 0;
exports.isWindow = isWindow;
exports.prefixToSelector = void 0;
var _dom = require("../dom");
function isWindow(node) {
  return node.toString() === '[object Window]';
}
function getWindow(node) {
  if (!node) {
    return window;
  }
  if (!isWindow(node)) {
    const {
      ownerDocument
    } = node;
    return ownerDocument ? ownerDocument.defaultView || window : window;
  }
  return node;
}
function isNonNullable(v) {
  if (v !== undefined && v !== null) {
    return true;
  }
  return false;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any

/**
 * @public
 */
function isFunction(x) {
  return typeof x === 'function';
}

/** @internal is the given object an Object? */
const isObject = obj => obj !== null && typeof obj === 'object';

/** @internal is the given object/value a promise? */
exports.isObject = isObject;
const isPromise = value => isObject(value) && isFunction(value.then);
exports.isPromise = isPromise;
function isReactInputElement(node) {
  return node.type === 'input';
}
function isReactHTMLElement(node) {
  return typeof node.type === 'string';
}
const isString = val => typeof val === 'string';

/**
 * Fork from https://github.com/juggle/resize-observer/blob/c92bcde23f23a5bc7b6d6e32449dec1e540c419f/src/ResizeObservation.ts#L6
 */
exports.isString = isString;
const isSVG = target => target instanceof SVGElement && 'getBBox' in target;
exports.isSVG = isSVG;
const isReplacedElement = target => {
  switch (target.tagName) {
    case 'INPUT':
      if (target.type !== 'image') {
        break;
      }
    case 'VIDEO':
    case 'AUDIO':
    case 'EMBED':
    case 'OBJECT':
    case 'CANVAS':
    case 'IFRAME':
    case 'IMG':
      return true;
    default:
  }
  return false;
};
exports.isReplacedElement = isReplacedElement;
const isNonReplacedElement = target => (0, _dom.isElement)(target) && !isSVG(target) && !isReplacedElement(target) && getComputedStyle(target).display === 'inline';

// 将 prefixCls='a b' 转换为 '.a .b'
exports.isNonReplacedElement = isNonReplacedElement;
const prefixToSelector = className => className.split(' ').map(cls => `.${cls}`).join(' ');
exports.prefixToSelector = prefixToSelector;
