// @universe-design/react-utilities/lib-commonjs/util/size.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultSizeEnum = exports.defaultExtraSizeEnum = void 0;
exports.getDefaultSizeAbbr = getDefaultSizeAbbr;
exports.getSizeAbbr = getSizeAbbr;
exports.isValidDefaultSize = isValidDefaultSize;
exports.isValidSize = isValidSize;
exports.sizeMap = exports.sizeEnum = void 0;
const sizeEnum = exports.sizeEnum = ['xl', 'lg', 'md', 'sm', 'xs'];
/**
 * @public
 */
const defaultSizeEnum = exports.defaultSizeEnum = ['large', 'medium', 'small'];

/**
 * 预设的一些尺寸
 * @public
 */

const defaultExtraSizeEnum = exports.defaultExtraSizeEnum = ['extra-large', 'extra-small'
// 'default',
];
/**
 * @public
 */

const sizeMap = exports.sizeMap = {
  'extra-large': 'xl',
  large: 'lg',
  medium: 'md',
  // default: 'md',
  small: 'sm',
  'extra-small': 'xs'
};
function isValidDefaultSize(sizeName) {
  return defaultSizeEnum.includes(sizeName);
}
function isValidSize(sizeName) {
  return defaultSizeEnum.includes(sizeName) || defaultExtraSizeEnum.includes(sizeName);
}
function getDefaultSizeAbbr(sizeName, defaultName) {
  if (isValidDefaultSize(sizeName)) {
    return sizeMap[sizeName];
  }
  if (defaultName && isValidDefaultSize(sizeName)) {
    return sizeMap[defaultName];
  }
  return '';
}
/**
 * @public
 */
function getSizeAbbr(sizeName) {
  if (isValidSize(sizeName)) {
    return sizeMap[sizeName];
  }
  return undefined;
}
