// @universe-design/react/lib/avatar/Avatar.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Avatar = void 0;
exports.sizeNormalize = sizeNormalize;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _clsx = _interopRequireDefault(require("clsx"));
var React = _interopRequireWildcard(require("react"));
var _reactUtilities = require("@universe-design/react-utilities");
var _hooks = require("../hooks");
var _configProvider = require("../config-provider");
var _util = require("../util");
var _AvatarGroupContext = require("./AvatarGroupContext");
var _common = require("./common");
var _tooltip = require("../tooltip");
var _composeRefs = _interopRequireDefault(require("../util/composeRefs"));
var _squircle = require("./squircle");
require("./style");
var _utils = require("./utils");
/**
 * @privateMark
 * Avatar
 * 圆形+空心（群组头像，内外 border 都行，现在其实都是做成图了）（加载态没有 border）
 * 圆形+实心（个人头像，无 border，icon + 图片）（加载态没有 border）
 * 方形+实心（应用头像，内 border，icon + 图片 + 字）（加载态没有 border）
 *
 * Avatar.Mask
 * 头像蒙层，出现在文字下方，图片上方
 *
 * Avatar.Group
 * 重叠头像，需要 mask clip 镂空，中间 2px 的间隙
 * 并排头像，直接并排就行
 *
 */

/**
 * @public
 */

const defaultSizeMap = {
  ['extra-small']: 20,
  ['small']: 24,
  ['medium']: 32,
  ['large']: 40,
  ['extra-large']: 48,
  ['default']: 32
};
function sizeNormalize(sizeProp) {
  let size = 0;
  if (typeof sizeProp === 'number') {
    size = sizeProp;
  } else if (!sizeProp || !defaultSizeMap[sizeProp]) {
    size = defaultSizeMap.default;
  } else {
    size = defaultSizeMap[sizeProp];
  }
  return size;
}

/**
 * @public
 */

const Avatar = exports.Avatar = /*#__PURE__*/React.forwardRef(
// eslint-disable-next-line max-lines-per-function
(props, ref) => {
  const {
    size: ctxSize,
    shape: ctxShape
  } = (0, _AvatarGroupContext.useAvatarGroupContext)();
  const {
    src,
    children,
    color = 'neutral',
    colorStyle = 'filled',
    gap = 0,
    className,
    style,
    icon,
    iconRenderer: iconRendererProps,
    srcSet,
    draggable,
    onError,
    onLoad,
    alt,
    tooltip,
    autoScaleContent = true,
    imgProps,
    shape = ctxShape,
    size: sizeProp = ctxSize,
    ...restProps
  } = props;
  const size = sizeNormalize(sizeProp);
  const [scale, setScale] = React.useState(1);
  const cssScope = (0, _common.useAvatarCssScope)();
  const slicedText = [...(typeof children === 'string' ? children : '')].reduce((pre, cur) => {
    const charStep = _util.RegexChinese.test(cur) || _util.RegexJapanese.test(cur) ? 2 : 1;
    if (pre.length + charStep <= 8) {
      if (pre.dist2lastLn + charStep > 4) {
        pre.result += `\n${cur}`;
        pre.dist2lastLn = charStep;
      } else {
        pre.result += cur;
        pre.dist2lastLn += charStep;
      }
      pre.length += charStep;
    }
    return pre;
  }, {
    result: '',
    dist2lastLn: 0,
    length: 0
  }).result;
  const [isImgExist, setIsImgExist] = React.useState(true);
  React.useEffect(() => {
    setIsImgExist(true);
  }, [src]);
  const containerRef = React.useRef(null);
  const [containerSizeRef, containerSize] = (0, _hooks.useSize)();
  React.useImperativeHandle(ref, () => containerRef.current);
  const textRef = (0, _reactUtilities.useResizeObserver)(entry => {
    var _containerRef$current, _containerRef$current2;
    const {
      contentRect: {
        width: textWidth,
        height: textHeight
      }
    } = entry;
    /**
     * 原型的相切矩形最大宽度
     * CW^2 = TW_max^2 + TH_max^2 = TW_max^2 * (1 + (TH / TW)^2)
     * TW_max = CW / SQRT(1 + (TH / TW)^2)
     */
    const containerWidth = (_containerRef$current = (_containerRef$current2 = containerRef.current) === null || _containerRef$current2 === void 0 ? void 0 : _containerRef$current2.clientHeight) !== null && _containerRef$current !== void 0 ? _containerRef$current : 0;
    const textWidthMax = (containerWidth - gap * 2) / Math.sqrt(1 + (textHeight / textWidth) ** 2);
    if (textWidth > textWidthMax) {
      setScale(textWidthMax / textWidth);
    }
  }, {
    disabled: !autoScaleContent
  });
  let content;
  let isImageAvatar = false;
  let isTextAvatar = false;
  if (typeof src === 'string' && isImgExist) {
    isImageAvatar = true;
    const {
      className: imgClassNameProp,
      ...restImgProps
    } = imgProps !== null && imgProps !== void 0 ? imgProps : {};
    content = /*#__PURE__*/React.createElement("img", (0, _extends2.default)({
      className: (0, _clsx.default)(cssScope('__image'), imgClassNameProp),
      src: src,
      draggable: draggable,
      srcSet: srcSet,
      alt: alt,
      onLoad: onLoad,
      onError: ev => {
        const errorFlag = (0, _util.isFunction)(onError) ? onError(ev) : undefined;
        if (errorFlag !== false) {
          setIsImgExist(false);
        }
      }
    }, restImgProps));
  } else if (src && isImgExist) {
    isImageAvatar = true;
    content = src;
  } else if (icon || iconRendererProps) {
    // icon 不为空的话是 icon avatar

    content = /*#__PURE__*/React.createElement("span", {
      className: cssScope('__icon')
    }, (iconRendererProps || (() => icon))({
      width: containerSize.width / 2,
      height: containerSize.height / 2
    }));
  } else {
    isTextAvatar = true;
    // slicedText 不为空的话是 text avatar
    content = /*#__PURE__*/React.createElement("span", {
      style: {
        // eslint-disable-next-line @typescript-eslint/no-magic-numbers
        fontSize: autoScaleContent ? 18 : (0, _utils.getTextAvatarFontSizeBySize)(size),
        transform: `translate(-50%, -50%) scale(${scale})`
      },
      className: cssScope('__text'),
      ref: textRef
    }, typeof children === 'string' ? slicedText : children);
  }
  const mergedRef = React.useMemo(() => isTextAvatar ? (0, _composeRefs.default)(containerRef, containerSizeRef) : containerRef, [containerSizeRef, isTextAvatar]);
  let wrapperStyle = {
    width: size,
    height: size
  };
  const mergedClassName = (0, _clsx.default)(cssScope(), {
    [cssScope(`-${color}`)]: !isImageAvatar && color,
    [cssScope(`-${colorStyle}`)]: shape === 'circle' ? colorStyle : colorStyle === 'filled',
    [cssScope(`-${shape}`)]: shape,
    [cssScope(`-${shape}-${size}`)]: shape === 'squircle',
    // [cssScope('-bordered')]: bordered,
    [cssScope('-image')]: isImageAvatar
  }, className);
  if (shape === 'square') {
    wrapperStyle = {
      ...wrapperStyle
    };
  }
  const avatar = /*#__PURE__*/React.createElement("span", (0, _extends2.default)({
    ref: mergedRef,
    style: {
      ...wrapperStyle,
      ...style
    },
    className: mergedClassName
  }, restProps), content, shape === 'squircle' && /*#__PURE__*/React.createElement(_squircle.SquircleAvatar, {
    bordered: isImageAvatar || color !== 'neutral',
    size: size
  }));
  if (tooltip) {
    return /*#__PURE__*/React.createElement(_tooltip.Tooltip, tooltip, avatar);
  }
  return avatar;
});
if (process.env.NODE_ENV !== "production") {
  Avatar.displayName = 'Avatar';
}
Object.assign(Avatar, {
  [_common.UD_AVATAR_ID]: true,
  [_configProvider.COMPONENT_VERSION_ID]: "3.28.3"
});
