// @universe-design/react/lib/avatar/AvatarGroup.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AvatarGroup = void 0;
var _u12o6wmt = _interopRequireDefault(require("./../.stylex/u12o6wmt"));
var _u1443mj = _interopRequireDefault(require("./../.stylex/u1443mj0"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _clsx = _interopRequireDefault(require("clsx"));
var React = _interopRequireWildcard(require("react"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _configProvider = require("../config-provider");
var _util = require("../util");
var _Avatar = require("./Avatar");
var _AvatarGroupContext = require("./AvatarGroupContext");
var _common = require("./common");
var _utils = require("./utils");
var _AvatarGroupMask = require("./AvatarGroupMask");
var _stylex2 = require("../stylex");
require("./style");
//.u1k6fduh{cursor:pointer;}
//.u1xqy1su{user-select:none;}
/**
 * Avatar.Group 的属性
 * @public
 */
const AvatarGroup = function (_ref) {
  let {
    maxCount = 5,
    children,
    avatars,
    shape = 'circle',
    size: sizeProp,
    spacing: spacingProp,
    maxProps = {},
    className,
    style,
    maxRenderer
  } = _ref;
  const avatarCssScope = (0, _common.useAvatarCssScope)();
  const size = (0, _Avatar.sizeNormalize)(sizeProp);
  const spacing = spacingProp !== null && spacingProp !== void 0 ? spacingProp : -size / 4;
  const cssScope = (0, _common.useAvatarGroupCssScope)();
  const cls = (0, _clsx.default)(cssScope(), className, spacing < 0 && cssScope('-stacked'));
  const hollowOutCls = (0, _AvatarGroupMask.useGroupSvgMaskCls)(size, spacing);
  let childNodes = [];
  if (children) {
    childNodes = (0, _util.toFlatArray)(children);
  } else if (avatars) {
    childNodes = avatars.map(avatar =>
    /*#__PURE__*/
    // eslint-disable-next-line react/jsx-key
    React.createElement(_Avatar.Avatar, avatar));
  }
  function getAvatarsWithMargin(nodes) {
    return nodes.map((child, index) => {
      if (/*#__PURE__*/React.isValidElement(child)) {
        var _child$key;
        return /*#__PURE__*/React.cloneElement(child, {
          key: (_child$key = child.key) !== null && _child$key !== void 0 ? _child$key : `avatar-key-${index}`,
          style: {
            ...child.props.style,
            marginLeft: index === 0 ? undefined : spacing
          }
        });
      }
      return child;
    });
  }
  if (childNodes.length > maxCount) {
    const children2Show = childNodes.slice(0, maxCount - 1);
    const children2Hide = childNodes.slice(maxCount - 1, childNodes.length);
    const {
      style: maxStyle,
      className: maxClassName,
      ...restMaxProps
    } = maxProps;
    const maxAvatarNode = /*#__PURE__*/React.createElement(_Avatar.Avatar, (0, _extends2.default)({
      key: "avatar-key-max",
      style: {
        marginLeft: spacing,
        ...maxStyle
      },
      shape: shape,
      size: size,
      className: (0, _clsx.default)(avatarCssScope('-max'), maxClassName, (0, _stylex.default)({
        "props": {
          "Bceei9c": _u12o6wmt.default,
          "famaaq": _u1443mj.default
        }
      }, _stylex2.cssFontMedium)),
      autoScaleContent: false
    }, restMaxProps), (0, _utils.getAvatarGroupMoreContent)({
      size,
      hideCount: children2Hide.length
    }));
    const maxNode = (0, _util.isFunction)(maxRenderer) ? maxRenderer({
      maxAvatar: maxAvatarNode,
      omittedAvatars: children2Hide.map((node, index) => ({
        node,
        index: index + children2Show.length
      }))
    }) : maxAvatarNode;
    childNodes = [...getAvatarsWithMargin(children2Show), maxNode];
  } else {
    childNodes = getAvatarsWithMargin(childNodes);
  }
  childNodes.forEach((child, index) => {
    if (/*#__PURE__*/React.isValidElement(child) && index < childNodes.length - 1) {
      childNodes[index] = /*#__PURE__*/React.cloneElement(child, {
        key: child.key,
        className: (0, _clsx.default)(child.props.className, hollowOutCls)
      });
    }
  });
  return /*#__PURE__*/React.createElement(_AvatarGroupContext.AvatarGroupContextProvider, {
    value: {
      shape,
      size
    }
  }, /*#__PURE__*/React.createElement("div", {
    className: cls,
    style: style
  }, childNodes));
};
exports.AvatarGroup = AvatarGroup;
Object.assign(AvatarGroup, {
  [_configProvider.COMPONENT_VERSION_ID]: "3.28.3"
});
