// @universe-design/react-ai/lib-commonjs/onboarding-animation/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AIOnboardingAnimation_unstable = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _avatarOnboardingAnimation = _interopRequireDefault(require("./avatar-onboarding-animation.json"));
var _reactUtilities = require("@universe-design/react-utilities");
var _LottieAnimation = _interopRequireDefault(require("../shared/components/LottieAnimation"));
/* eslint-disable @typescript-eslint/no-magic-numbers */

/**
 * AI onboarding 动画组件属性
 *
 * @beta
 */

const Segments = {
  intro: [0, 147],
  defaultToAvatar: [146, 177],
  avatarToDefault: [177, 146],
  confirmAvatar: [176, 251],
  confirmDefault: [252, 351]
};
const animationGraph = [{
  from: 'none',
  to: 'default',
  segment: Segments.intro
}, {
  from: 'default',
  to: 'avatar',
  segment: Segments.defaultToAvatar
}, {
  from: 'avatar',
  to: 'default',
  segment: Segments.avatarToDefault
}, {
  from: 'avatar',
  to: 'avatarConfirmed',
  segment: Segments.confirmAvatar
}, {
  from: 'default',
  to: 'defaultConfirmed',
  segment: Segments.confirmDefault
}];

/**
 * AI onboarding 动画组件属性
 *
 * @beta
 */

/**
 * AI onboarding 动画组件
 *
 * @beta
 */
const AIOnboardingAnimation_unstable = exports.AIOnboardingAnimation_unstable = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    animationRef,
    ...htmlProps
  } = _ref;
  const lottieAnimationRef = React.useRef(null);
  const currentStage = React.useRef('none');
  function gotoStage(target) {
    const availablePath = animationGraph.find(edge => edge.from === currentStage.current && edge.to === target);
    if (availablePath) {
      var _lottieAnimationRef$c;
      (_lottieAnimationRef$c = lottieAnimationRef.current) === null || _lottieAnimationRef$c === void 0 ? void 0 : _lottieAnimationRef$c.playSegments([...availablePath.segment], currentStage.current === 'none');
      currentStage.current = target;
    }
    process.env.NODE_ENV !== 'production' && (process.env.NODE_ENV !== "production" ? (0, _reactUtilities.warning)(!availablePath, (0, _reactUtilities.devWarning)('OnboardingAnimation', `can't convert stage '${currentStage.current}' to stage '${target}'`)) : void 0);
  }
  React.useImperativeHandle(animationRef, () => ({
    toDefault: () => {
      gotoStage('default');
    },
    toAvatar: () => {
      gotoStage('avatar');
    },
    confirm: () => {
      if (currentStage.current === 'avatar') {
        gotoStage('avatarConfirmed');
      } else {
        gotoStage('defaultConfirmed');
      }
    }
  }), []);
  return /*#__PURE__*/React.createElement(_LottieAnimation.default, (0, _extends2.default)({
    ref: ref,
    animationData: _avatarOnboardingAnimation.default,
    lottieAnimationRef: lottieAnimationRef
  }, htmlProps));
});
