// @universe-design/react/lib/shared/components/trigger/StreamlineArrow/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StreamlineArrow = void 0;
var _react = _interopRequireDefault(require("react"));
/* eslint-disable max-len */

const downPath = 'M4.438 1.993L7.253 5.16a1 1 0 001.494 0l2.815-3.166A5.938 5.938 0 0116 0H0c1.696 0 3.311.725 4.438 1.993z';
const downStrokePath = 'M8-.5H0v1c1.553 0 3.033.664 4.065 1.825l2.814 3.166a1.5 1.5 0 002.242 0l2.814-3.166A5.438 5.438 0 0116 .5v-1H8z';
const rightPath = 'M1.993 11.562L5.16 8.747a1 1 0 000-1.494L1.993 4.438A5.938 5.938 0 010 0v16c0-1.696.725-3.311 1.993-4.438z';
const rightStrokePath = 'M-.5 8v8h1c0-1.553.664-3.033 1.825-4.065l3.166-2.814a1.5 1.5 0 000-2.242L2.325 4.065A5.438 5.438 0 01.5 0h-1v8z';
const StreamlineArrow = _ref => {
  let {
    className,
    style,
    direction = 'horizontal',
    hasStroke = false
  } = _ref;
  const a = 16;
  const b = 8;
  const offset = 0;
  const d = (() => {
    if (direction === 'horizontal') {
      if (hasStroke) {
        return downStrokePath;
      }
      return downPath;
    }
    if (hasStroke) {
      return rightStrokePath;
    }
    return rightPath;
  })();
  return /*#__PURE__*/_react.default.createElement("svg", {
    width: direction === 'horizontal' ? a : b,
    height: direction === 'horizontal' ? b : a,
    viewBox: direction === 'horizontal' ? `0 ${offset} ${a} ${b + offset}` : `${offset} 0 ${b + offset} ${a}`,
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg",
    className: className,
    style: style
  }, /*#__PURE__*/_react.default.createElement("path", {
    className: `${className}__content`,
    d: d
  }));
};
exports.StreamlineArrow = StreamlineArrow;
