// @byted-nodex/metrics/lib/collector/cursor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SlowCursor = exports.Cursor = void 0;
const measure_1 = require("../measure");
const reservoir_1 = require("../reservoir");
const metrics_codec_1 = require("@byted-nodex/metrics-codec");
class Cursor {
    constructor(options) {
        this.values = new reservoir_1.Queue();
        this.index = options.index;
        this.metric = options.metric;
    }
    get isExpired() {
        return !this.values.size;
    }
    get active() {
        return this.values.active;
    }
    emit(value, ...values) {
        this.walkValues(value);
        if (values.length) {
            for (const v of values) {
                this.walkValues(v);
            }
        }
    }
    walkValues(value) {
        if (this.metric.forceMultiFields) {
            // 在多值环境下，要求 field 与预定义的 field 一致
            const type = value.type === metrics_codec_1.MetricType.Meter ? metrics_codec_1.MetricType.Counter : value.type;
            const field = this.metric.fields.get(type, value.suffix);
            if (!field) {
                const existsFields = this.metric.fields.originFields;
                throw new Error(`NOT found this Field(${value.suffix}, ${metrics_codec_1.MetricType[value.type]}) in the fields: ${JSON.stringify(existsFields === null || existsFields === void 0 ? void 0 : existsFields.map(f => ({ name: f.name, type: metrics_codec_1.MetricType[f.type] })))}`);
            }
        }
        let rateValue;
        let field;
        switch (value.type) {
            case metrics_codec_1.MetricType.Meter:
                // meter 就是一个 counter 和一个加了 .rate 后缀的 rateCounter
                // 所以这里生成个 rateCounter 搞里头，再改 value.type 为 Counter
                rateValue = (0, measure_1.incr)(value.value);
                if (value.suffix) {
                    rateValue.suffix = `${value.suffix}.rate`;
                }
                else {
                    rateValue.suffix = 'rate';
                }
                this.walkValues(rateValue);
                value.type = metrics_codec_1.MetricType.Counter;
                break;
            case metrics_codec_1.MetricType.Histogram:
                // 处理 buckets
                field = this.metric.fields.get(value.type, value.suffix);
                if (!field) {
                    return this.metric.logger.error('didn\'t found the buckets for histogram metric %s. It will ignore this value"', value.suffix);
                }
                value.buckets = field.buckets;
                break;
            default:
                break;
        }
        value.policy = this.metric.valuePolicy;
        this.values.mergeOrInsert(value);
    }
    pack(message, lastMessage) {
        if (!this.values.active) {
            return 0;
        }
        if (this.metric.forceMultiFields) {
            return this.forcePackMultiFields(message, lastMessage);
        }
        else {
            return this.compatPack(message, lastMessage);
        }
    }
    gc(opt) {
        this.values.gc(opt);
        return this.values.size === 0;
    }
    getTags() {
        const sortedGlobalTagPos = this.metric.sortedGlobalTagPos;
        const sortedUserTagPos = this.metric.sortedUserTagPos;
        const { globalTags, sortedTags, allTagsLen } = this.metric;
        const tags = new Array(allTagsLen);
        sortedGlobalTagPos.forEach((tagIndex, originIndex) => {
            tags[tagIndex] = globalTags[originIndex];
        });
        sortedUserTagPos.forEach((tagIndex, originIndex) => {
            tags[tagIndex] = { key: sortedTags[originIndex], value: this.index[originIndex] };
        });
        return tags;
    }
    prepackTags() {
        if (!this.metric.cacheTagBytes) {
            return (0, metrics_codec_1.encodeTagBytes)(this.getTags());
        }
        if (!this.cacheTagBytes) {
            this.cacheTagBytes = (0, metrics_codec_1.encodeTagBytes)(this.getTags());
        }
        return this.cacheTagBytes;
    }
    /** 强制多值 */
    forcePackMultiFields(message, lastMessage) {
        const codecFields = [];
        const lastCodecFields = [];
        const values = [];
        const lastValues = [];
        const fields = this.metric.fields;
        // 遍历链表，将所有记录序列化进 message
        this.values.marshall((vs, node) => {
            var _a;
            const field = fields.get(node.type, node.suffix);
            const inCodecFields = ((_a = field === null || field === void 0 ? void 0 : field.meta) === null || _a === void 0 ? void 0 : _a.codecFields) || [];
            vs.forEach(v => {
                if (v.lastValues) {
                    lastValues.push(...v.lastValues);
                    lastCodecFields.push(...inCodecFields);
                }
                values.push(...v.values);
            });
            codecFields.push(...inCodecFields);
        });
        // 木有值
        if (!values.length && !lastValues.length) {
            return 0;
        }
        let packedMetricCount = 0;
        const tagBytes = this.prepackTags();
        // 上个周期的值
        if (lastMessage && lastValues.length) {
            packedMetricCount++;
            this.appendMetric(lastMessage, metrics_codec_1.MetricType.MultiFields, this.metric.name, lastCodecFields, tagBytes, lastValues);
        }
        // 当前值
        if (values.length) {
            packedMetricCount++;
            this.appendMetric(message, metrics_codec_1.MetricType.MultiFields, this.metric.name, codecFields, tagBytes, values);
        }
        return packedMetricCount;
    }
    /** 兼容模式打包，存在多值单值混合 */
    compatPack(message, lastMessage) {
        let packedMetricCount = 0;
        const fields = this.metric.fields;
        const tagBytes = this.prepackTags();
        this.values.marshall((vs, node) => {
            vs.forEach(v => {
                let codecFields;
                // timer 有可能是多值
                if (node.type === metrics_codec_1.MetricType.Timer) {
                    if (!node.policy.compatTimer) {
                        codecFields = fields.getDefaultTimerFields();
                    }
                }
                else if (node.type === metrics_codec_1.MetricType.Histogram) {
                    codecFields = fields.getDefaultHistogramFields(v.suffix);
                }
                else if (node.type === metrics_codec_1.MetricType.DeltaCounterType) {
                    codecFields = fields.getDefaultDeltaFields();
                }
                const reportName = v.suffix ? `${this.metric.name}.${v.suffix}` : this.metric.name;
                // 上个周期的值
                if (lastMessage && v.lastValues) {
                    packedMetricCount++;
                    this.appendMetric(lastMessage, v.type, reportName, codecFields, tagBytes, v.lastValues);
                }
                // 当前周期的值
                packedMetricCount++;
                this.appendMetric(message, v.type, reportName, codecFields, tagBytes, v.values);
            });
        });
        return packedMetricCount;
    }
    appendMetric(m, type, name, fields, tagBytes, values) {
        const reportFields = fields || [];
        m.addMetric(type, name, reportFields, [{ tags: tagBytes, values }]);
        if (this.metric.clientOps.debug) {
            this.logReportMetric(name, type, values, reportFields, m);
        }
    }
    getDebugTags() {
        return this.getTags()
            .map(t => `${t.key}=${t.value}`)
            .join('|');
    }
    logReportMetric(name, type, values, fields, m) {
        const tags = this.getDebugTags();
        const debugValues = fields.length
            ? fields.map((f, i) => `${f.name}=${values[i]}`).join('|')
            : values.join('|');
        // debug 模式打印 metric 信息
        this.metric.logger.info('Report metric: %s, type: %s, tags: %s, values: %s, timestamp: %s', name, metrics_codec_1.MetricType[type], tags, debugValues, new Date(Number(m.header.timestamp) * 1000).toLocaleTimeString());
    }
}
exports.Cursor = Cursor;
class SlowCursor extends Cursor {
    constructor(options) {
        super(options);
        this.tagNames = options.tagNames;
    }
    getTags() {
        const { globalTags } = this.metric;
        const userTags = this.tagNames.map((t, index) => ({
            key: t,
            value: this.index[index],
        }));
        return globalTags.concat(userTags).sort((a, b) => (a.key < b.key ? -1 : a.key > b.key ? 1 : 0));
    }
}
exports.SlowCursor = SlowCursor;
