// @byted-nodex/metrics/lib/collector/dict.js
"use strict";
/**
 * 用字典来索引序列
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DictCollector = void 0;
const collector_1 = require("./collector");
const cursor_1 = require("./cursor");
const consts_1 = require("../consts");
const cursorSymbol = Symbol.for('cursor');
class DictCollector extends collector_1.Collector {
    constructor(metric) {
        super(metric);
        this.metric = metric;
        this.maxCursorSize = consts_1.defaultMaxCursorSize;
        this.sortedTags = metric.sortedTags;
        this.dict = {};
    }
    get(opt) {
        let cursorContainer = this.dict;
        for (const tagIndex of opt.index) {
            cursorContainer[tagIndex] = cursorContainer[tagIndex] || {};
            cursorContainer = cursorContainer[tagIndex];
        }
        if (!cursorContainer[cursorSymbol]) {
            if (this.size >= this.maxCursorSize) {
                this.metric.logger.error('Reach max size of cursor: %s/%s (tagNames: [ %s ])', this.size, this.maxCursorSize, this.sortedTags.join(','));
                return null;
            }
            this.size++;
            return (cursorContainer[cursorSymbol] = new cursor_1.Cursor({
                index: opt.index,
                metric: this.metric,
            }));
        }
        return cursorContainer[cursorSymbol];
    }
    gc(opt) {
        const depth = this.sortedTags.length;
        const traverse = (node, currDepth = 1, check) => {
            const nodeList = Object.keys(node);
            const delNode = (v) => {
                delete node[v];
                if (check)
                    check();
            };
            nodeList.forEach(v => {
                if (currDepth < depth) {
                    return traverse(node[v], currDepth + 1, () => {
                        // 如果当前节点没有子元素了，清理节点
                        if (!Object.keys(node[v]).length)
                            delNode(v);
                    });
                }
                const cursor = node[v][cursorSymbol];
                if (cursor.gc(opt)) {
                    this.size--;
                    delNode(v);
                }
            });
        };
        traverse(this.dict);
    }
    each(cb) {
        const depth = this.sortedTags.length;
        const traverse = (node, currDepth = 1) => {
            for (const v in node) {
                if (currDepth === depth) {
                    cb(node[v][cursorSymbol]);
                }
                else {
                    traverse(node[v], currDepth + 1);
                }
            }
        };
        traverse(this.dict);
    }
}
exports.DictCollector = DictCollector;
