// @universe-design/react/es/pro/timezone-picker/const.js
import moment from '@lark/moment-timezone';
import keyBy from 'lodash/keyBy';
import once from 'lodash/once';
import { TIMEZONE_RAW_META_LIST } from './timezoneMeta';
export { TIMEZONE_LINK_MAP } from './timezoneLink';
const getTimezoneLocaleKey = timezone => {
  const lastSlashPos = timezone.lastIndexOf('/');
  return timezone.slice(lastSlashPos + 1);
};
export const extendRawMeta = rawMeta => ({
  ...rawMeta,
  l: `Tz_${getTimezoneLocaleKey(rawMeta.v)}`,
  s: Object.entries(rawMeta.c).map(_ref => {
    let [key, value] = _ref;
    return Object.assign([key, ...value], {
      c: `Ct_${key}`
    });
  })
});
export const TIMEZONE_META_LIST = TIMEZONE_RAW_META_LIST.map(extendRawMeta);
const timezoneSorter = (a, b) => {
  const offsetA = moment().tz(a.v).utcOffset();
  const offsetB = moment().tz(b.v).utcOffset();
  return offsetA - offsetB;
};

/** 数组按照 GMT 偏移量从小到大排序 */
export const getSortedTimezoneMetaList = once(() => [...TIMEZONE_META_LIST].sort(timezoneSorter));

/** 获取 timezone meta map */
export const getTimezoneMetaMap = once(() => keyBy(TIMEZONE_META_LIST, item => item.v));
