// @universe-design/react/es/pro/timezone-picker/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import * as React from 'react';
import moment from '@lark/moment-timezone';
import isEmpty from 'lodash/isEmpty';
import { useCssScope, useLocale } from '../../config-provider/hooks';
import { Select } from '../../select';
import { Text } from '../../text';
import { useConstant } from '@universe-design/react-utilities';
import { Highlighter } from '../../shared/components/highlighter';
import { useSelectListItemCssScope } from '../../select/common';
import { searchTimezoneMetaList, fillInTimezoneMetaList, getTimezoneMeta, useTimezoneLocale, arrayJoin, useMetaTransformer, splitKeywords } from './utils';
import './style';
export const TimezonePicker = /*#__PURE__*/React.memo(props => {
  const defaultMetaTransformer = useMetaTransformer();
  const {
    value: valueProp,
    defaultValue: defaultValueProp,
    date,
    metaTransformer = defaultMetaTransformer,
    ...restProps
  } = props;
  const dateString = date === null || date === void 0 ? void 0 : date.toString();

  /**
   * Context
   */
  const cssScope = useSelectListItemCssScope();
  const timezonePickerCssScope = useCssScope('timezone-picker');

  /**
   * Locale
   */
  const locale = useLocale();
  const timezoneLocale = useTimezoneLocale();

  /**
   * Value
   */
  const value = React.useMemo(() => {
    var _getTimezoneMeta$v, _getTimezoneMeta;
    return (_getTimezoneMeta$v = (_getTimezoneMeta = getTimezoneMeta(valueProp)) === null || _getTimezoneMeta === void 0 ? void 0 : _getTimezoneMeta.v) !== null && _getTimezoneMeta$v !== void 0 ? _getTimezoneMeta$v : valueProp;
  }, [valueProp]);

  // 线上逻辑，兼容时区不完整
  const defaultValue = useConstant(() => {
    var _getTimezoneMeta$v2, _getTimezoneMeta2;
    return (_getTimezoneMeta$v2 = (_getTimezoneMeta2 = getTimezoneMeta(defaultValueProp)) === null || _getTimezoneMeta2 === void 0 ? void 0 : _getTimezoneMeta2.v) !== null && _getTimezoneMeta$v2 !== void 0 ? _getTimezoneMeta$v2 : defaultValueProp;
  });

  /**
   * Search
   */
  const [searchValue, setSearchValue] = React.useState('');
  const handleSearch = React.useCallback(inputValue => {
    setSearchValue(inputValue);
  }, []);

  /**
   * Options
   */

  // 原始 meta list 的 text 和 search 字段只是占位符，需要做一下「fill in」操作
  const filledTimezoneMetaList = React.useMemo(() => {
    const _date = dateString ? moment(dateString) : undefined;
    return timezoneLocale ? fillInTimezoneMetaList(timezoneLocale, _date, metaTransformer) : [];
  }, [timezoneLocale, dateString, metaTransformer]);
  const options = React.useMemo(() => {
    // TODO 之后把 matcher 能力和 scorer 能力也放进去，label 直接基于 matches 数组去做渲染就行了
    const filteredTimezoneMetaList = searchTimezoneMetaList(filledTimezoneMetaList, searchValue);
    const searchWords = splitKeywords(searchValue);
    const highlightClassName = cssScope('__content-highlight');
    const _options = filteredTimezoneMetaList.map(timezoneMeta => {
      var _timezoneMeta$cities;
      return {
        value: timezoneMeta.v,
        title: timezoneMeta.l,
        label: /*#__PURE__*/React.createElement("div", {
          className: timezonePickerCssScope('__label')
        }, /*#__PURE__*/React.createElement(Highlighter, {
          searchWords: searchWords,
          textToHighlight: timezoneMeta.l,
          markRenderer: text => /*#__PURE__*/React.createElement("span", {
            className: highlightClassName
          }, text)
        }), isEmpty(timezoneMeta.cities) ? null : /*#__PURE__*/React.createElement(Text, {
          as: "div",
          className: timezonePickerCssScope('__subtitle'),
          lineClamp: 2
        }, /*#__PURE__*/React.createElement("span", null, timezoneLocale === null || timezoneLocale === void 0 ? void 0 : timezoneLocale.Ct_Includes), arrayJoin((_timezoneMeta$cities = timezoneMeta.cities) === null || _timezoneMeta$cities === void 0 ? void 0 : _timezoneMeta$cities.map((city, index) => /*#__PURE__*/React.createElement("span", {
          key: `item_${index}`,
          className: highlightClassName
        }, city)), i => /*#__PURE__*/React.createElement("span", {
          key: `joiner_${i}`
        }, "\u3001"))))
      };
    });
    return _options;
  }, [timezoneLocale, filledTimezoneMetaList, searchValue, cssScope, timezonePickerCssScope]);
  return /*#__PURE__*/React.createElement(Select
  // 由于这里是手动控制 options，所以用 noOptionsContent 而不是 notFoundContent
  , _extends({
    noOptionsContent: locale.global.notFoundContent
  }, restProps, 'value' in props ? {
    value
  } : {}, {
    defaultValue,
    filterOption: false,
    searchValue,
    onSearch: handleSearch,
    options
  }));
});
export { chinaMainlandMetaTransformer, getTimezoneNames, getTimezoneFormatData } from './utils';
export { TimezoneLabel } from './TimezoneLabel';
