// @universe-design/react/es/pro/timezone-picker/utils.js
import { getGMTTimezone } from '@lark/moment-timezone';
import compact from 'lodash/compact';
import countBy from 'lodash/countBy';
import isEmpty from 'lodash/isEmpty';
import keys from 'lodash/keys';
import memoize from 'lodash/memoize';
import { useContextSelector } from 'use-context-selector';
import { getTimezoneMetaMap, TIMEZONE_LINK_MAP, TIMEZONE_META_LIST, getSortedTimezoneMetaList } from './const';
import { ConfigContext } from '../../config-provider';
import { useImportLocale } from '../../config-provider/hooks/useImportLocale';
import enUS from './locales/en-US';

// 提取gmt偏移量
const GMT_STRING_REGEXP = /\(?(gmt|utc)?\s*(\+?|-)\s*([\d:]+)([^\d:])*/i;
function normalizeGMTOffset() {
  let gmtString = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
  return gmtString.replace(GMT_STRING_REGEXP, ($0, $1, $2, offset) => $2 + offset.replace(/(\d{2})/g, '$1:').replace(/(:\B)/g, ''));
}

/**
 * 包含 compatible 逻辑的获取 timezone meta 函数
 * @param value - timezone 值
 * @returns timezone meta值
 */
export const getTimezoneMeta = value => {
  if (!value) {
    return void 0;
  }
  const timezoneMetaMap = getTimezoneMetaMap();
  return timezoneMetaMap[value] || timezoneMetaMap[TIMEZONE_LINK_MAP[value]];
};
// 预期搜索 “+8” 也能匹配 “+08:00” 的选项
export function splitKeywords(keywords) {
  return keywords.replace(/\+(\d)(?!\d)/g, '+ $1').split(/\s+/g);
}
export const searchTimezoneMetaList = (timezoneMetaList, keywords) => {
  if (!keywords) {
    return timezoneMetaList;
  }
  return compact(timezoneMetaList.map(item => {
    const _text = item.l.toLowerCase();
    const _keywords = keywords.toLowerCase();
    const matchedSearchEntries = item.s.filter(entry => {
      var _entry$c;
      const [enPure, zhCnPure, zhCnPinyinFull, zhCnPinyinInitials = '', zhCnPinyinIndexes = []] = entry;
      if ((_entry$c = entry.c) !== null && _entry$c !== void 0 && _entry$c.toLowerCase().includes(_keywords)) {
        return true;
      }
      if (zhCnPure.includes(keywords)) {
        return true;
      }
      if (enPure.toLowerCase().includes(_keywords)) {
        return true;
      }
      if (zhCnPinyinInitials.includes(keywords)) {
        return true;
      }
      if (zhCnPinyinIndexes.includes(zhCnPinyinFull.replace(/\s/g, '').indexOf(keywords))) {
        return true;
      }
      return false;
    });
    if (matchedSearchEntries.length > 0) {
      return {
        ...item,
        cities: matchedSearchEntries.map(entry => entry.c)
      };
    }
    if (_text.includes(_keywords)) {
      return item;
    }
    const words = splitKeywords(_keywords);
    if (words.every(w => _text.includes(w))) {
      return item;
    }
    const gmtOffset = normalizeGMTOffset(_text);
    if (gmtOffset.includes(normalizeGMTOffset(keywords))) {
      return item;
    }
    return null;
  }, []));
};
const extendSearchEntries = (searchEntries, localeData) => searchEntries.map(entry => {
  // order: [enPure, zhCnPure, zhCnPinyinFull]
  const zhCnPinyinFull = entry[2];
  const zhCnPinyinFullList = zhCnPinyinFull.split(/\s+/g);
  const zhCnPinyinInitials = zhCnPinyinFullList.map(w => w[0]).join('');
  const zhCnPinyinIndexes = zhCnPinyinFullList.reduce((res, curr) => {
    res.push(res[res.length - 1] + curr.length);
    return res;
  }, [0]);
  const extendedEntry = Object.assign([entry[0], entry[1], entry[2], zhCnPinyinInitials, zhCnPinyinIndexes], {
    c: localeData[entry.c]
  });
  return extendedEntry;
});

/**
 * 默认格式化函数
 */
export const defaultFormatter = p => p ? `(${p.GMTOffset}) ${p.zoneName}` : '';
const getTimezoneNameCount = memoize(localeData => countBy(TIMEZONE_META_LIST, item => localeData[item.l]));
getTimezoneNameCount.cache = new WeakMap();

/**
 * timezone 格式化
 * @internal
 * @experimental
 * @param value - timezone 值
 * @param localeData - timezone 国际化文案数据
 * @returns 格式化数据
 */
export const getTimezoneFormatData = function (value, localeData) {
  var _options$transformer;
  let options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
  const meta = getTimezoneMeta(value);
  if (!meta) {
    return void 0;
  }
  const localeKey = ((_options$transformer = options.transformer) === null || _options$transformer === void 0 ? void 0 : _options$transformer.call(options, meta, getTimezoneMetaMap()).l) || meta.l;
  const timeName = localeData[localeKey];
  const isUnique = getTimezoneNameCount(localeData)[timeName] === 1;
  const cityLocales = keys(options.cities).map(city => localeData[`Ct_${city}`]);
  return {
    GMTOffset: getGMTTimezone(value, {
      padNumber: true,
      date: options.date
    }),
    timeName,
    zoneName: `${timeName}${isUnique || isEmpty(cityLocales) ? '' : ` - ${cityLocales[0]}`}`,
    isUnique,
    cities: cityLocales
  };
};
export const fillInTimezoneMetaList = (localeData, date, transformer) => {
  const _sortedTimezoneMetaList = getSortedTimezoneMetaList();
  const _timezoneList = _sortedTimezoneMetaList.map(item => {
    const {
      v: value,
      c: cities,
      s: searchEntries
    } = item;
    const temp = {
      v: value,
      c: cities,
      l: defaultFormatter(getTimezoneFormatData(value, localeData, {
        date,
        cities,
        transformer
      })),
      s: extendSearchEntries(searchEntries, localeData)
    };
    return temp;
  });
  return _timezoneList;
};

/**
 * @public
 * 获取所有时区的名称值
 */
export const getTimezoneNames = () => TIMEZONE_META_LIST.map(o => o.v);
export const arrayJoin = (array, joiner) => {
  if (!array) {
    return [];
  }
  const result = [];
  for (let i = 0; i < array.length; ++i) {
    result.push(array[i]);
    if (i !== array.length - 1) {
      result.push(joiner(i));
    }
  }
  return result;
};
export const importTimezoneLocale = async locale => {
  let promise = import(`./locales/${locale}.js`);
  return promise.then(_ref => {
    let {
      default: localeData
    } = _ref;
    return localeData;
  });
};
export const useTimezoneLocale = () => useImportLocale({
  importLocale: importTimezoneLocale,
  fallbackLocale: enUS
});
export const useMetaTransformer = () => useContextSelector(ConfigContext, o => {
  var _o$timezone;
  return (_o$timezone = o.timezone) === null || _o$timezone === void 0 ? void 0 : _o$timezone.metaTransformer;
});
export const chinaMainlandMetaTransformer = (meta, metaMap) => {
  if (meta.v === 'Asia/Hong_Kong' || meta.v === 'Asia/Macau' || meta.v === 'Asia/Taipei') {
    return {
      ...meta,
      l: metaMap['Asia/Shanghai'].l
    };
  }
  return meta;
};
