// @universe-design/react/lib/dropdown/Button.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DropdownButton = void 0;
var _react = _interopRequireDefault(require("react"));
var _button = require("../button");
var _Dropdown = require("./Dropdown");
var _MoreOutlined = _interopRequireDefault(require("@universe-design/icons-react/MoreOutlined"));
var _hooks = require("../config-provider/hooks");
var _clsx = _interopRequireDefault(require("clsx"));
require("./style");
/**
 * @public
 * @docgen
 * <Dropdown.Button> 的属性和方法。
 */

const DropdownButton = props => {
  const {
    type,
    disabled,
    htmlType,
    children,
    overlay,
    trigger,
    align,
    visible,
    defaultVisible,
    placement = 'bottomRight',
    href,
    icon = /*#__PURE__*/_react.default.createElement(_MoreOutlined.default, null),
    title,
    className,
    style,
    getPopupContainer,
    onVisibleChange,
    buttonsRender,
    onClick,
    size,
    ...restProps
  } = props;
  const cssScope = (0, _hooks.useCssScope)('dropdown-button');
  const dropdownProps = {
    align,
    overlay,
    disabled,
    visible,
    defaultVisible,
    placement,
    trigger,
    onVisibleChange,
    getPopupContainer,
    ...restProps
  };
  const leftButton = /*#__PURE__*/_react.default.createElement(_button.Button, {
    size: size,
    type: type,
    disabled: disabled,
    onClick: onClick,
    htmlType: htmlType,
    href: href,
    title: title
  }, children);
  const rightButton = /*#__PURE__*/_react.default.createElement(_button.Button, {
    size: size,
    disabled: disabled,
    type: type,
    shape: "square",
    icon: icon
  });
  const [leftButtonToRender, rightButtonToRender] = buttonsRender ? buttonsRender([leftButton, rightButton]) : [leftButton, rightButton];
  return /*#__PURE__*/_react.default.createElement("div", {
    style: style,
    className: (0, _clsx.default)(cssScope(), className)
  }, leftButtonToRender, /*#__PURE__*/_react.default.createElement(_Dropdown.Dropdown, dropdownProps, rightButtonToRender));
};
exports.DropdownButton = DropdownButton;
