// @universe-design/react/lib/dropdown/Dropdown.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Dropdown = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _hooks = require("../config-provider/hooks");
var _usePopperSidecar = require("../config-provider/hooks/usePopperSidecar");
var _util = require("../util");
var _DropdownMenuContext = require("./DropdownMenuContext");
var _composeRefs = _interopRequireDefault(require("../util/composeRefs"));
var _SlideUpDownTransition = require("../shared/components/transition/SlideUpDownTransition");
require("./style");
/* eslint-disable max-lines-per-function */

/**
 * @public
 */

/**
 * @public
 * @docgen
 */

// eslint-disable-next-line max-lines-per-function
const Dropdown = exports.Dropdown = /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
  const {
    prefixCls: prefixClsInProps,
    align: alignInProps,
    placement = 'bottomLeft',
    showAction,
    hideAction,
    overlayClassName,
    overlayStyle,
    trigger = 'hover',
    disabled = false,
    onOverlayClick,
    overlay,
    children,
    udTriggerRef: udTriggerRefProp,
    ...otherProps
  } = props;
  const cssScope = (0, _hooks.useCssScope)('dropdown');
  const Popper = (0, _usePopperSidecar.usePopperSidecar)();
  const prefixCls = prefixClsInProps || cssScope();
  const udTriggerRef = (0, _react.useRef)(null);
  const mergedUdTriggerRef = _react.default.useMemo(() => (0, _composeRefs.default)(udTriggerRef, udTriggerRefProp), [udTriggerRef, udTriggerRefProp]);
  const overlayNode = (0, _react.useMemo)(() => (0, _util.isFunction)(overlay) ? overlay() : overlay, [overlay]);
  const onClick = (0, _react.useCallback)(e => {
    var _udTriggerRef$current;
    (0, _util.isFunction)(onOverlayClick) && onOverlayClick(e);
    if (e && 'domEvent' in e && e.domEvent.isDefaultPrevented()) {
      return;
    }
    (_udTriggerRef$current = udTriggerRef.current) === null || _udTriggerRef$current === void 0 ? void 0 : _udTriggerRef$current.setVisible(false);
  }, [onOverlayClick]);
  const renderOverlay = /*#__PURE__*/_react.default.createElement(_DropdownMenuContext.DropdownMenuContext.Provider, {
    value: (0, _react.useMemo)(() => ({
      className: `${prefixCls}__overlay`,
      onClick
    }), [onClick, prefixCls])
  }, overlayNode);
  const alignPoint = trigger.indexOf('contextMenu') !== -1;
  return /*#__PURE__*/_react.default.createElement(Popper, (0, _extends2.default)({}, otherProps, {
    disabled: disabled,
    alignPoint: alignPoint,
    prefixCls: prefixCls,
    ref: ref,
    udTriggerRef: mergedUdTriggerRef,
    popupClassName: overlayClassName,
    popupStyle: overlayStyle,
    action: trigger,
    showAction: showAction,
    hideAction: hideAction,
    popupPlacement: placement,
    popupAlign: alignInProps,
    popupMotion: "slide-up",
    popup: renderOverlay,
    showArrow: false,
    arrowHasStroke: false,
    Transition: _SlideUpDownTransition.SlideUpDownTransition,
    defaultMainAxisOffsetWithoutArrow: 4,
    __screenshotAlwaysOpenEnabled: true
  }), _react.default.useCallback(_ref => {
    let {
      visible
    } = _ref;
    return 'selected' in children.props ? children : /*#__PURE__*/_react.default.cloneElement(children, {
      selected: visible
    });
  }, [children]));
});
if (process.env.NODE_ENV !== "production") {
  Dropdown.displayName = 'Dropdown';
}
