// @universe-design/react/es/react-mini-form/internal/useFormList.js
import { useRef, useMemo, useCallback } from 'react';
import { asArray, normalizeValidator, marshalNamePath } from '../shared';
import get from 'lodash/get';
import cloneDeep from 'lodash/cloneDeep';
import { useFormInternalContextSelector } from './FormInternalContext';
import { useFieldInternal } from './useFieldInternal';
import isEqual from 'lodash/isEqual';
import { useLatest } from '../../hooks';

/**
 * fork from formik
 * @see https://github.com/jaredpalmer/formik/blob/b9cc2536a1edb9f2d69c4cd20ecf4fa0f8059ade/packages/formik/src/FieldArray.tsx#L108
 **/
const copyArrayLike = arrayLike => {
  if (!arrayLike) {
    return [];
  } else if (Array.isArray(arrayLike)) {
    return [...arrayLike];
  } else {
    const maxIndex = Object.keys(arrayLike).map(key => parseInt(key)).reduce((max, el) => el > max ? el : max, 0);
    return Array.from({
      ...arrayLike,
      length: maxIndex + 1
    });
  }
};
function usePersistValue(path) {
  const valueRef = useRef();
  return useFormInternalContextSelector(ctx => {
    valueRef.current = get(ctx, path);
    return valueRef;
  });
}

/** Form List 配置 */

/** FormList 返回行数据(fields) */

/** FormList 操作帮助方法 */

/**
 * @docgen
 * @public
 * @typeParam FieldValue - 字段值
 */

function getGhostFormListName(name) {
  return `__ud-form-list-${name}`;
}
export function useFormList(options) {
  const {
    name: listName,
    rowKey,
    validate
  } = options;
  /** 获取数组型 path，为了方面后续直接使用 `path.concat` 来组合 name */
  const listNameMultiLevel = asArray(listName);
  const _setFieldValue = useFormInternalContextSelector(ctx => ctx._setFieldValue);
  const setFieldError = useFormInternalContextSelector(ctx => ctx.setFieldError);
  const _setFieldTouchedState = useFormInternalContextSelector(ctx => ctx._setFieldTouchedState);
  const _setFieldTouched = useFormInternalContextSelector(ctx => ctx._setFieldTouched);
  const touched = useFormInternalContextSelector(ctx => ctx.touched);

  /** 校验相关 */
  const valuesRef = usePersistValue(['values', ...listNameMultiLevel]);
  const ghostFormListName = getGhostFormListName(listName);
  const {
    error,
    validateField
  } = useFieldInternal({
    name: ghostFormListName,
    validate: function () {
      // 标准化 validator
      const validators = asArray(validate).map(normalizeValidator);
      // 然后注入第一个参数为当前表单 list 的值
      return validators.map(validator => ({
        ...validator,
        validate(_value, namePath) {
          // value proxy: ghost list name -> list name
          return validator.validate(valuesRef.current, namePath);
        }
      }));
    }()
  });
  const validateRef = useLatest(validate);
  const triggerListValidate = useCallback(() => {
    if (validateRef.current) {
      setTimeout(() => {
        // 触发的时候，需要手动再设置一下 touched
        _setFieldTouched(marshalNamePath(ghostFormListName), true);
        validateField();
      });
    }
  }, [validateRef, validateField, ghostFormListName, _setFieldTouched]);

  /** 校验相关 END */

  const setTouchedTimer = useRef();
  const lastTouchedSnapshotRef = useRef();
  const getCurrentValue = useCallback((rootValues, path) => {
    const currentValue = get(rootValues, path);
    const currentValueList = copyArrayLike(currentValue);
    return currentValueList;
  }, []);
  const updateFields = useCallback(function (updateOption) {
    let config = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    // actions 默认加上 triggerByUser = true，使其可以触发 Form 的 `onValuesChange` 事件
    // 需要注意一点，如果多次调用 actions 操作，onValuesChange 也会多次触发
    const {
      triggerByUser = true,
      ...othersConfig
    } = config;
    const {
      transformFn,
      transformErrorFn = transformFn,
      transformTouchedFn = transformFn
    } = updateOption;
    if (transformFn) {
      _setFieldValue(marshalNamePath(listNameMultiLevel), values => {
        const currentValue = getCurrentValue(values, listNameMultiLevel);
        const resultValue = transformFn(currentValue);
        // 触发校验
        triggerListValidate();
        return resultValue;
      }, {
        triggerByUser,
        ...othersConfig
      });
    }
    if (transformErrorFn) {
      setFieldError(listNameMultiLevel, errors => {
        const currentError = getCurrentValue(errors, listNameMultiLevel);
        const resultError = transformErrorFn(currentError);
        return resultError;
      });
    }
    if (transformTouchedFn) {
      /**
       * touched 本身在 formItem unregister 的时候，会被设置成 false，从而会导致 touched 混乱
       * 所以使用 Snapshot 在 1 tick 之后来还原 touched，但是这种情况对于连续使用多个 actions 会造成问题
       * 引入 timer 是为了支持连续使用多个 actions，将他们的 touched 逻辑 batch 在一起
       */
      const lastTouchedSnapshot = lastTouchedSnapshotRef.current;
      const currentTouched = getCurrentValue(touched, listNameMultiLevel);
      const currentTouchedWillTransform = setTouchedTimer.current ? lastTouchedSnapshot : currentTouched;
      const resultTouched = transformTouchedFn(currentTouchedWillTransform || currentTouched);
      const resultTouchedSnapshot = cloneDeep(resultTouched);
      lastTouchedSnapshotRef.current = resultTouchedSnapshot;
      if (setTouchedTimer.current) {
        clearTimeout(setTouchedTimer.current);
      }
      setTouchedTimer.current = setTimeout(() => {
        setTouchedTimer.current = undefined;
        _setFieldTouchedState(marshalNamePath(listNameMultiLevel), resultTouchedSnapshot);
      });
    }
  }, [listNameMultiLevel, _setFieldTouchedState, getCurrentValue, setFieldError, _setFieldValue, touched, triggerListValidate]);
  const actions = useMemo(() => ({
    remove(index) {
      updateFields({
        transformFn: list => {
          const res = [...list];
          res.splice(index, 1);
          return res;
        }
      });
    },
    push() {
      for (var _len = arguments.length, appendItems = new Array(_len), _key = 0; _key < _len; _key++) {
        appendItems[_key] = arguments[_key];
      }
      updateFields({
        transformFn: list => [...list, ...appendItems],
        transformTouchedFn: false,
        transformErrorFn: false
      });
    },
    unshift() {
      for (var _len2 = arguments.length, prependItems = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
        prependItems[_key2] = arguments[_key2];
      }
      updateFields({
        transformFn: list => [...prependItems, ...list],
        transformTouchedFn: list => [...prependItems.map(() => undefined), ...list],
        transformErrorFn: list => [...prependItems.map(() => undefined), ...list]
      });
    },
    insert(index) {
      for (var _len3 = arguments.length, insertItems = new Array(_len3 > 1 ? _len3 - 1 : 0), _key3 = 1; _key3 < _len3; _key3++) {
        insertItems[_key3 - 1] = arguments[_key3];
      }
      updateFields({
        transformFn: list => {
          const res = [...list];
          res.splice(index, 0, ...insertItems);
          return res;
        },
        transformTouchedFn: list => {
          const res = [...list];
          res.splice(index, 0, ...insertItems.map(() => undefined));
          return res;
        },
        transformErrorFn: list => {
          const res = [...list];
          res.splice(index, 0, ...insertItems.map(() => undefined));
          return res;
        }
      });
    },
    move(from, to) {
      updateFields({
        transformFn: list => {
          const res = [...list];
          const [item] = res.splice(from, 1);
          res.splice(to, 0, null);
          res[to] = item;
          return res;
        }
      });
    },
    swap(swapItemIndex1, swapItemIndex2) {
      updateFields({
        transformFn: list => {
          const res = [...list];
          const [swapItem1, swapItem2] = [res[swapItemIndex1], res[swapItemIndex2]];
          res[swapItemIndex1] = swapItem2;
          res[swapItemIndex2] = swapItem1;
          return res;
        }
      });
    }
  }), [updateFields]);
  function getFields(fieldsValue) {
    if (Array.isArray(fieldsValue)) {
      return fieldsValue.map((item, index) => ({
        name: listNameMultiLevel.concat(`${index}`),
        // 如果有key，优先取key
        // 默认取 index
        key: rowKey ? get(item, rowKey) : index,
        remove() {
          // TODO: 此处如果多次调用会出问题，可能需要根据 `rowKey` 来识别删除会好一点？
          actions.remove(index);
        }
      }));
    }
    return [];
  }

  // 决定是否触发刷新的 Key
  function getFieldsUpdateKey(fieldList) {
    return fieldList.map(field => [field.key, field.name]);
  }
  const prevFieldsRef = useRef([]);
  const fields = useFormInternalContextSelector(ctx => {
    // 减少刷新次数
    const currentFields = getFields(get(ctx.values, listNameMultiLevel));
    if (isEqual(getFieldsUpdateKey(prevFieldsRef.current), getFieldsUpdateKey(currentFields))) {
      return prevFieldsRef.current;
    }
    prevFieldsRef.current = currentFields;
    return prevFieldsRef.current;
  });
  return {
    fields,
    actions,
    error,
    validateField
  };
}
