// @universe-design/react/es/react-mini-form/internal/useFieldInternal.js
import get from 'lodash/get';
import * as React from 'react';
import { useConstant, useForceUpdate } from '@universe-design/react-utilities';
import { useLatest } from '../../hooks';
import { marshalNamePath, asArray, normalizeValidator, DEFAULT_VALIDATE_TRIGGER } from '../shared';
import { useFormInternalContextSelector } from './FormInternalContext';

/**
 * @public
 * @docgen
 */

/**
 * @public
 */

/**
 * @public
 * @docgen
 */

/**
 * useFieldContext
 */
export function useFieldContext(_ref) {
  let {
    name
  } = _ref;
  const multiLevelPath = asArray(name);
  return {
    value: useFormInternalContextSelector(v => get(v.values, multiLevelPath)),
    error: useFormInternalContextSelector(v => v.errors ? get(v.errors, multiLevelPath) : undefined)
  };
}
let count = 0;
const useFieldID = () => useConstant(() => `field-${count++}`);

/**
 * useField
 */
export function useFieldInternal(props) {
  const {
    name,
    validate,
    valuePropName,
    validateTrigger,
    trigger = 'onChange'
  } = props;
  const fieldID = useFieldID();
  const namePathStr = marshalNamePath(name);
  const {
    value,
    error
  } = useFieldContext({
    name
  });
  const _setFieldValue = useFormInternalContextSelector(v => v._setFieldValue);
  const _registerField = useFormInternalContextSelector(v => v._registerField);
  const _unregisterField = useFormInternalContextSelector(v => v._unregisterField);
  const _setFieldTouched = useFormInternalContextSelector(v => v._setFieldTouched);
  const validate$ = useFormInternalContextSelector(v => v.validate$);
  const forceUpdate = useForceUpdate();

  // validate 统一为 Validator[]
  // 1、当 Validator 没有设置 validateTrigger，会使用 Field 的 validateTrigger
  // 2、在 1 条件下，当 field 没有设置validateTrigger，会使用默认的 validateTrigger
  const mergedValidators = (() => {
    const validators = asArray(validate).map(normalizeValidator);
    validators.forEach(v => {
      if (!v.validateTrigger.length) {
        v.validateTrigger = asArray(validateTrigger);
      }
      if (!v.validateTrigger.length) {
        v.validateTrigger.push(DEFAULT_VALIDATE_TRIGGER);
      }
    });
    return validators;
  })();
  const mergedValidatorsRef = useLatest(mergedValidators);
  const allValidateTriggers = (() => {
    const triggerSet = new Set();
    mergedValidators.forEach(v => {
      v.validateTrigger.forEach(t => {
        if (t !== DEFAULT_VALIDATE_TRIGGER) {
          triggerSet.add(t);
        }
      });
    });
    return [...triggerSet];
  })();
  React.useEffect(() => {
    _registerField({
      fieldID,
      serializedName: namePathStr,
      validate: mergedValidatorsRef
    });
    return () => {
      _unregisterField(fieldID);
    };
  }, [_registerField, _unregisterField, mergedValidatorsRef, fieldID, namePathStr]);

  /**
   * Super mega handle Change
   */
  const handleChangeMega = React.useCallback(eventOrValue => {
    /**
     * 此处 forceUpdate 是因为用了 use-context-selector 导致 change 之后 Field 没有重新渲染，
     * 导致 child 拿到旧值，引发光标位置的抖动
     * issue： https://code.byted.org/universe-ui/react/issues/937
     */
    forceUpdate();
    // First arg is event
    let _value;
    if (typeof eventOrValue === 'object' && 'nativeEvent' in eventOrValue && eventOrValue.nativeEvent instanceof Event) {
      if (eventOrValue.target instanceof HTMLInputElement) {
        switch (eventOrValue.target.type) {
          case 'radio':
          case 'checkbox':
            _value = eventOrValue.target[valuePropName !== null && valuePropName !== void 0 ? valuePropName : 'checked'];
            break;
          case 'file':
            throw new Error('Field cannot be used on file inputs since they are read-only. Use a onChange listener instead.');
          default:
            _value = eventOrValue.target[valuePropName !== null && valuePropName !== void 0 ? valuePropName : 'value'];
        }
      } else if (eventOrValue.target instanceof HTMLSelectElement) {
        if (!eventOrValue.target.multiple) {
          _value = eventOrValue.target.value;
        } else {
          const res = [];
          for (let i = 0; i < eventOrValue.target.options.length; i++) {
            const opt = eventOrValue.target.options[i];
            if (opt.selected) {
              res.push(opt.value);
            }
          }
          _value = res;
        }
      } else {
        // radio group
        switch (eventOrValue.target.type) {
          case 'radio':
          case 'checkbox':
            _value = eventOrValue.target[valuePropName !== null && valuePropName !== void 0 ? valuePropName : 'checked'];
            break;
          default:
            _value = eventOrValue.target[valuePropName !== null && valuePropName !== void 0 ? valuePropName : 'value'];
            break;
        }
      }
    } else {
      _value = eventOrValue;
    }
    _setFieldTouched(namePathStr, true);
    _setFieldValue(namePathStr, _value, {
      triggerByUser: true,
      triggerName: trigger,
      shouldValidate: true
    });
  }, [_setFieldTouched, _setFieldValue, forceUpdate, namePathStr, trigger, valuePropName]);
  const validateField = React.useCallback(options => {
    validate$.next({
      serializedName: namePathStr,
      triggerName: (options === null || options === void 0 ? void 0 : options.triggerName) || DEFAULT_VALIDATE_TRIGGER
    });
  }, [namePathStr, validate$]);
  return {
    value,
    handleChange: handleChangeMega,
    error,
    allValidateTriggers,
    serializedName: namePathStr,
    trigger,
    validateField
  };
}
