// @byted-service/zti-helper/lib/consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CMD = exports.DPS_AGENT_USER = exports.DPS_AGENT_PSM = exports.DPS_AGENT_ENABLED = exports.DPS_AGENT_SOCKET_PATH = exports.DPS_AGENT_ENABLE_PATH = exports.DPS_AGENT_PSMENV = exports.DPS_AGENT_USER_ENV = exports.SCHEMA_ID = exports.SCHEMA_VDC = exports.SCHEMA_R = exports.SCHEMA_NS = exports.FETCH_TOKEN_FROM_PATH_FAIL = exports.TOKEN_PATH_NOT_EXIST = exports.QUIT_POLLING = exports.GET_TOKEN_FROM_AGENT_FAIL = exports.SYNC_TOKEN_WITH_FALLBACK = exports.GET_TOKEN_FROM_DPS_AGENT = exports.GET_TOKEN_FROM_STRING = exports.GET_TOKEN_SUCCEED = exports.GET_TOKEN_EXPIRED = exports.GET_TOKEN_FAIL = exports.ZTI_ENV = exports.UPDATE_INTERVAL_SECS = exports.THRESHOLD_HOUR = exports.DEFAULT_CONTEXT_TIMEOUT = exports.DEFAULT_TOKEN_POLL_INTERVAL = exports.TOKEN_STR = exports.TOKEN_PATH = exports.DEFAULT_AUDIENCE = exports.AGENT_SOCKET_PATH = exports.ZTI_AGENT_SOCKET_PATH = exports.IS_DEBUG = exports.IS_JEST = void 0;
const tslib_1 = require("tslib");
const path_1 = tslib_1.__importDefault(require("path"));
// for jest
exports.IS_JEST = process.env.IS_NODE_JS_ZTI_HELPER_JEST === 'true';
// for debug
exports.IS_DEBUG = process.env.IS_NODE_JS_ZTI_HELPER_DEBUG === 'true';
exports.ZTI_AGENT_SOCKET_PATH = '/var/run/zti-agent.sock';
exports.AGENT_SOCKET_PATH = exports.IS_JEST
    ? path_1.default.resolve(__dirname, '../test/zti-agent.sock')
    : exports.ZTI_AGENT_SOCKET_PATH;
exports.DEFAULT_AUDIENCE = 'zti';
exports.TOKEN_PATH = process.env.SEC_TOKEN_PATH;
exports.TOKEN_STR = process.env.SEC_TOKEN_STRING;
exports.DEFAULT_TOKEN_POLL_INTERVAL = 1000 * 10;
exports.DEFAULT_CONTEXT_TIMEOUT = 1000 * 5;
exports.THRESHOLD_HOUR = 1000 * 60 * 60 * 8;
exports.UPDATE_INTERVAL_SECS = 180;
exports.ZTI_ENV = 'ZTI_ENV';
exports.GET_TOKEN_FAIL = 'get_token_error';
exports.GET_TOKEN_EXPIRED = 'get_token_expired';
exports.GET_TOKEN_SUCCEED = 'get_token';
exports.GET_TOKEN_FROM_STRING = 'get_token_from_string';
exports.GET_TOKEN_FROM_DPS_AGENT = 'get_token_from_dps_agent';
exports.SYNC_TOKEN_WITH_FALLBACK = 'get_token_with_fallback';
exports.GET_TOKEN_FROM_AGENT_FAIL = 'get_token_from_agent_fail';
exports.QUIT_POLLING = 'quit_polling';
exports.TOKEN_PATH_NOT_EXIST = 'token_path_not_exist';
exports.FETCH_TOKEN_FROM_PATH_FAIL = 'fetch_token_path_error';
exports.SCHEMA_NS = 'ns';
exports.SCHEMA_R = 'r';
exports.SCHEMA_VDC = 'vdc';
exports.SCHEMA_ID = 'id';
exports.DPS_AGENT_USER_ENV = 'INFSEC_SEC_USER';
exports.DPS_AGENT_PSMENV = 'INFSEC_SEC_PSM';
exports.DPS_AGENT_ENABLE_PATH = 'GET_SEC_TOKEN_STRING_FROM_DAEMON';
exports.DPS_AGENT_SOCKET_PATH = '/opt/tmp/sock/.unix_sock_agent_1234567890.sock';
exports.DPS_AGENT_ENABLED = process.env[exports.DPS_AGENT_ENABLE_PATH];
exports.DPS_AGENT_PSM = process.env[exports.DPS_AGENT_PSMENV];
exports.DPS_AGENT_USER = process.env[exports.DPS_AGENT_USER_ENV];
exports.CMD = Buffer.from(JSON.stringify({
    PSM: exports.DPS_AGENT_PSM,
    Cmd: 1,
    User: exports.DPS_AGENT_USER,
}) + '\n');
