// @byted-service/zti-helper/lib/helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZtiJWTHelper = exports.symbolForZtiJWTHelperInstance = void 0;
const tslib_1 = require("tslib");
const grpc = tslib_1.__importStar(require("@grpc/grpc-js"));
const utils = tslib_1.__importStar(require("util"));
const jwt = tslib_1.__importStar(require("jsonwebtoken"));
const consts = tslib_1.__importStar(require("./consts"));
const workload_grpc_pb_1 = require("./gen/workload_grpc_pb");
const fs_1 = tslib_1.__importDefault(require("fs"));
const metrics_1 = tslib_1.__importDefault(require("./metrics"));
const workload_pb_1 = require("./gen/workload_pb");
const lodash_clonedeep_1 = tslib_1.__importDefault(require("lodash.clonedeep"));
const utill_1 = require("./utill");
const logger_1 = require("./logger");
const singleflight_1 = tslib_1.__importDefault(require("@byted-service/singleflight"));
const logger_2 = tslib_1.__importDefault(require("@byted-service/logger"));
// import net from 'net';
const readFileAsync = utils.promisify(fs_1.default.readFile);
const statAsync = utils.promisify(fs_1.default.stat);
const initZtiAgentKey = '$_init_zti_agent_key_$';
exports.symbolForZtiJWTHelperInstance = Symbol('$_symbol_for_zti_jwt_helper_single_instance_$');
class ZtiJWTHelper {
    constructor(options = {}) {
        this.sf = new singleflight_1.default();
        this.updateTokenFromPathTime = 0;
        if (global[exports.symbolForZtiJWTHelperInstance]) {
            return global[exports.symbolForZtiJWTHelperInstance];
        }
        global[exports.symbolForZtiJWTHelperInstance] = this;
        this.logger =
            options.logger ||
                new logger_2.default({ prefix: '[zti-helper]', level: 'info' });
        if (!utill_1.IS_SCM) {
            this.sf.do(initZtiAgentKey, async () => {
                try {
                    await this._createZtiAgent();
                }
                catch (error) /* istanbul ignore next */ {
                    logger_1.debugLogger.warn('connect to zti agent fail, %s', String(error));
                }
            });
        }
        this.logger.info('zti-helper instance is created');
    }
    async getJwtSVID(forceUpdate) {
        // 1st priority source process.env.SEC_TOKEN_STRIN without rotation
        if (process.env.SEC_TOKEN_STRING) {
            (0, metrics_1.default)(consts.GET_TOKEN_FROM_STRING, {});
            return process.env.SEC_TOKEN_STRING;
        }
        // 2st priority source SEC_TOKEN_PATH
        const now = Date.now() / 1000;
        if ((forceUpdate ||
            now - this.updateTokenFromPathTime >
                consts.UPDATE_INTERVAL_SECS) &&
            consts.TOKEN_PATH) {
            await this._fetchTokenFromPath(now);
        }
        if (this.tokenFromPath) {
            (0, metrics_1.default)(consts.GET_TOKEN_SUCCEED, {});
            return this.tokenFromPath;
        }
        // 3rd priority source zti-agent's socket path
        if (!this.ztiAgent) {
            try {
                // It will retry to create agent if it is failed to create agent
                await this.sf.do(initZtiAgentKey, async () => /* istanbul ignore next */ {
                    await this._createZtiAgent();
                });
            }
            catch (error) /* istanbul ignore next */ {
                (0, metrics_1.default)(consts.GET_TOKEN_FROM_AGENT_FAIL, {});
                throw error;
            }
        }
        if (this.ztiAgent) {
            return this._pollOnce();
        }
        else {
            (0, metrics_1.default)(consts.GET_TOKEN_FAIL, {});
            throw new Error('no valid jwt-svid token exist');
        }
    }
    decodeGDPRorJwtSVID(tokenString) {
        const token = jwt.decode(tokenString, { complete: true });
        if (!token) {
            throw new Error('decode fail token: ' + tokenString);
        }
        if (!token.header || !token.header.kid) {
            return (0, lodash_clonedeep_1.default)(token.payload);
        }
        else {
            return (0, utill_1.decodeIdentityFromTokenString)(tokenString, token).getLegacyIdentityFromZTI();
        }
    }
    decodeJwtSVID(tokenString) {
        return (0, utill_1.decodeIdentityFromTokenString)(tokenString);
    }
    close() {
        this.sf
            .do(initZtiAgentKey, async () => {
            // do nothing
        })
            .then(() => {
            if (this.ztiAgent) {
                this.ztiAgent.close();
            }
        });
    }
    async _pollOnce() {
        logger_1.debugLogger.info('fetching token from zti-agent ... ');
        try {
            const jwtSVID = await this._fetchJWTSVID({
                audience: consts.DEFAULT_AUDIENCE,
            });
            (0, metrics_1.default)(consts.GET_TOKEN_SUCCEED, {});
            return jwtSVID.token;
        }
        catch (error) {
            logger_1.debugLogger.error('fail to fetch jwt-svid from workload api, error: %s', String(error));
            (0, metrics_1.default)(consts.GET_TOKEN_FROM_AGENT_FAIL, {});
            throw error;
        }
    }
    async _createZtiAgent() {
        const spiffeWorkloadAPIClient = new workload_grpc_pb_1.SpiffeWorkloadAPIClient(consts.AGENT_SOCKET_PATH, grpc.credentials.createInsecure());
        await new Promise((resolve, reject) => {
            spiffeWorkloadAPIClient.waitForReady(Date.now() + 1000, err => {
                if (err) {
                    /* istanbul ignore next */
                    return reject(err);
                }
                resolve();
            });
        });
        this.ztiAgent = spiffeWorkloadAPIClient;
    }
    async _fetchTokenFromPath(now) {
        const tokenPath = consts.TOKEN_PATH || '';
        try {
            await statAsync(tokenPath);
        }
        catch (error) {
            (0, metrics_1.default)(consts.FETCH_TOKEN_FROM_PATH_FAIL, {});
            logger_1.debugLogger.error('SEC_TOKEN_PATH does not exist , error: %s', String(error));
            throw error;
        }
        try {
            const tokenString = await readFileAsync(tokenPath, {
                encoding: 'utf-8',
            });
            this.tokenFromPath = tokenString.toString().trim();
            this.updateTokenFromPathTime = now;
        }
        catch (error) /* istanbul ignore next */ {
            (0, metrics_1.default)(consts.FETCH_TOKEN_FROM_PATH_FAIL, {});
            logger_1.debugLogger.error('fail to fetch token from SEC_TOKEN_PATH，error: %o', error);
            throw error;
        }
    }
    async _fetchJWTSVID(params) {
        var _a;
        const audience = [params.audience, ...(params.extraAudiences || [])];
        const req = new workload_pb_1.JWTSVIDRequest();
        req.setAudienceList(audience);
        req.setSpiffeId(((_a = params.subject) === null || _a === void 0 ? void 0 : _a.string()) || '');
        const meta = new grpc.Metadata();
        meta.set('workload.spiffe.io', 'true');
        const resp = await new Promise((resolve, reject) => {
            this.ztiAgent.fetchJWTSVID(req, meta, (err, res) => {
                if (err) {
                    return reject(err);
                }
                resolve(res);
            });
        });
        if (!resp || !resp.getSvidsList() || resp.getSvidsList().length < 1) {
            throw new Error('there were no SVIDs in the response');
        }
        const token = resp.getSvidsList()[0].getSvid();
        return (0, utill_1.parseInsecure)(token, audience);
    }
}
exports.ZtiJWTHelper = ZtiJWTHelper;
