// @byted-service/zti-helper/lib/zeroTrustIdentity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZeroTrustIdentity = void 0;
const consts_1 = require("./consts");
class ZeroTrustIdentity {
    constructor(spiffeID, expiry) {
        this.spiffeID = spiffeID;
        this.expiry = expiry;
    }
    validateZTI() {
        if (this.spiffeID.isZero()) {
            throw new Error('the SPIFFE ID is the zero value');
        }
        if (!this.expiry) {
            throw new Error('the expiry is zero, we do not support zero expiry as token without expiration');
        }
        const paths = this.spiffeID.path.split('/');
        paths.forEach((path) => {
            const schema = path.split(':');
            if (schema.length !== 2) {
                return;
            }
            const [key, value] = schema;
            switch (key.toLowerCase()) {
                case consts_1.SCHEMA_NS:
                    this.namespace = value;
                    break;
                case consts_1.SCHEMA_R:
                    this.region = value;
                    break;
                case consts_1.SCHEMA_VDC:
                    this.VDC = value;
                    break;
                case consts_1.SCHEMA_ID:
                    this.ID = value;
                    break;
            }
        });
    }
    isValidated() {
        return this.ID.trim() !== '';
    }
    getLegacyIdentityFromZTI() {
        if (!this.isValidated()) {
            this.validateZTI();
        }
        const identity = {
            version: 0,
            authority: this.namespace,
            expireTime: this.expiry,
        };
        if (this.namespace.toLowerCase() === 'user') {
            identity.primaryAuthType = 'user';
            identity.user = this.ID;
        }
        else {
            identity.primaryAuthType = 'psm';
            identity.psm = this.ID;
        }
        return identity;
    }
}
exports.ZeroTrustIdentity = ZeroTrustIdentity;
