// @universe-design/react/src/_stories/common/CssVarDebugger.tsx
/** @jsxImportSource @emotion/react */
import React, { useLayoutEffect, useState } from 'react';
import { useImmer } from 'use-immer';
import List from '@universe-design/rc-virtual-list';
import UDBizTokens from '@universe-design/styles/esm/universe-design/index.light';
import { Input, Popover, useSize, ColorPicker, Drawer, useDerivedThemeModeState } from '../..';

const getBodyCssVar = () =>
  Object.keys(UDBizTokens)
    .map(k => ({
      k,
      v: getComputedStyle(document.documentElement).getPropertyValue(`--${k}`),
    }))
    .filter(o => o.v);

function escapeRegex(string: string) {
  return string.replace(/[-/\\^$*+?.()|[\]{}]/g, '\\$&');
}

function vscodeLikeSearch(content: string, inputValue: string | { raw: string; reg: RegExp }): { matched: boolean } {
  if (typeof inputValue === 'string') {
    const reg = new RegExp([...inputValue].map(escapeRegex).join('.*'), 'i');
    const matched = reg.test(content);
    return {
      matched,
    };
  }
  const matched = inputValue.reg.test(content);
  return {
    matched,
  };
}

const ColorSelect = ({
  size = 16,
  value,
  onChange,
}: {
  size?: number;
  value?: string;
  onChange?: (v: string | undefined) => any;
}) => {
  const [show, setShowState] = React.useState(false);
  if (!show) {
    return (
      <div
        style={{
          cursor: 'pointer',
          width: size,
          height: size,
          background: value,
        }}
        onMouseEnter={() => setShowState(true)}
      />
    );
  }
  return (
    <Popover
      trigger={['click']}
      mode="pure"
      content={<ColorPicker css={{ border: 'none' }} value={value} onChange={onChange} />}
    >
      <div
        style={{
          cursor: 'pointer',
          width: size,
          height: size,
          background: value,
        }}
      />
    </Popover>
  );
};

export const CssVarDebugger = () => {
  const [visible, setVisible] = useState(false);
  const [search, setSearch] = useState('');
  const [state, produce] = useImmer(getBodyCssVar);
  const [ref, { height }] = useSize<HTMLDivElement>();
  const mode = useDerivedThemeModeState();
  useLayoutEffect(() => {
    produce(draft => {
      draft.splice(0, draft.length, ...getBodyCssVar());
    });
  }, [mode, produce]);
  const handlerNode = (
    <div
      css={{
        cursor: 'pointer',
        position: 'fixed',
        borderTopLeftRadius: 4,
        borderBottomLeftRadius: 4,
        border: '2px solid var(--line-border-component)',
        top: '50%',
        transform: 'translateY(-50%)',
        right: 0,
        padding: 4,
        ':hover': {
          boxShadow: 'var(--shadow-md-left)',
        },
      }}
      onClick={() => setVisible(true)}
    >
      Token Editor
    </div>
  );
  return (
    <>
      {!visible ? handlerNode : null}
      <Drawer bodyStyle={{ padding: 0 }} visible={visible} onClose={() => setVisible(false)}>
        <div
          style={{
            height: '100vh',
            overflow: 'hidden',
            display: 'flex',
            flexDirection: 'column',
          }}
        >
          <div>
            <Input
              style={{
                borderRadius: 0,
              }}
              value={search}
              onChange={e => setSearch(e.target.value)}
            />
          </div>
          <div
            ref={ref}
            style={{
              height: 'calc(100vh - 32px)',
              overflow: 'hidden',
            }}
          >
            <List
              data={state.filter(o => vscodeLikeSearch(o.k, search).matched)}
              itemKey={o => o.k}
              virtual
              height={height}
              itemHeight={78}
            >
              {({ k, v }) => (
                <div
                  style={{
                    padding: '0 1rem',
                  }}
                >
                  <div style={{ padding: '1rem 0' }}>{k}</div>
                  <Input
                    size="small"
                    value={v}
                    onChange={e => {
                      produce(draft => {
                        const item = draft.find(o => o.k === k);
                        if (item) {
                          item.v = e.target.value;
                          document.documentElement.style.setProperty(`--${k}`, e.target.value);
                        }
                      });
                    }}
                    suffix={
                      <ColorSelect
                        value={v}
                        onChange={_v => {
                          produce(draft => {
                            const item = draft.find(o => o.k === k);
                            if (item) {
                              item.v = _v ?? '';
                              document.documentElement.style.setProperty(`--${k}`, _v ?? '');
                            }
                          });
                        }}
                      />
                    }
                  />
                </div>
              )}
            </List>
          </div>
        </div>
      </Drawer>
    </>
  );
};
