// @universe-design/react/src/_stories/common/PerformanceContent.tsx
/* eslint-disable @babel/new-cap */
import * as React from 'react';
import { useForceUpdate } from '@universe-design/react-utilities';
import { useEffectOnce, useUpdateEffect } from '../../hooks';
import { Space, Switch, Button } from '../..';

export const usePerformanceLogger = (componentName: string, ...rest: unknown[]) => {
  const renderStartAt = performance.now();

  useEffectOnce(() => {
    setTimeout(
      () => console.log(`${componentName} mounted, which costs ${performance.now() - renderStartAt} ms`, ...rest),
      0,
    );
    return () => console.log(`${componentName} unmounted`);
  });

  useUpdateEffect(() => {
    console.log(`${componentName} updated, which costs ${performance.now() - renderStartAt} ms`, ...rest);
  });
};

export const PerformanceLogger = ({ componentName, children }: React.PropsWithChildren<{ componentName: string }>) => {
  usePerformanceLogger(componentName);
  return <>{children}</>;
};

const defaultItemRenderer = (i: number) => <div>{i}</div>;

const PerformanceContentLogger: React.FC<{
  componentName: string;
  count?: number;
  itemRenderer?: (i: number) => React.ReactNode;
}> = ({ componentName, count = 1000, itemRenderer = defaultItemRenderer }) => {
  usePerformanceLogger(componentName);
  return (
    <div>
      {Array(count)
        .fill(0)
        .map((o, i) => (
          <React.Fragment key={i}>{itemRenderer(i)}</React.Fragment>
        ))}
    </div>
  );
};

export const PerformanceContent: React.FC<{
  componentName: string;
  count?: number;
  itemRenderer?: (i: number) => React.ReactNode;
}> = ({ componentName, ...restProps }) => {
  const forceUpdate = useForceUpdate();
  const [visible, setVisible] = React.useState(false);
  return (
    <>
      <Space>
        {`${componentName} 性能测试`}
        <Switch checked={visible} onChange={setVisible} />
        <Button onClick={forceUpdate}>更新重渲染</Button>
      </Space>
      <div>{visible ? <PerformanceContentLogger {...{ componentName, ...restProps }} /> : null}</div>
    </>
  );
};
