// @universe-design/react/src/_stories/common/WindowMock.tsx
import React from 'react';

const Dot: React.FC<{ style?: React.CSSProperties }> = ({ style }) => (
  <span
    style={{
      height: 12,
      width: 12,
      borderRadius: '50%',
      display: 'inline-block',
      marginRight: 8,
      ...style,
    }}
  />
);

export const WindowMock: React.FC<
  React.PropsWithChildren<{
    windowSize?: { width?: number; height?: number };
    scale?: number;
  }>
> = ({ children, windowSize, scale = 1 }) => (
  <div
    style={{
      borderRadius: 4,
      overflow: 'hidden',
      boxShadow: '#999 0 2px 4px',
      width: 'fit-content',
    }}
  >
    <div
      style={{
        padding: '0 12px',
        height: 40,
        display: 'flex',
        alignItems: 'center',
        backgroundColor: '#222',
      }}
    >
      <Dot style={{ backgroundColor: '#ED594A' }} />
      <Dot style={{ backgroundColor: '#FDD800' }} />
      <Dot style={{ backgroundColor: '#5AC05A' }} />
    </div>
    <div
      style={{
        position: 'relative',
        backgroundColor: '#fff',
        width: windowSize?.width ? windowSize.width / scale : 'auto',
        height: windowSize?.height ? windowSize.height / scale : 'auto',
        overflow: 'auto',
        zoom: scale,
      }}
    >
      {children}
    </div>
  </div>
);
