// @byted-service/zti-helper/lib/gen/workload_grpc_pb.js
// GENERATED CODE -- DO NOT EDIT!

'use strict';
var grpc = require('@grpc/grpc-js');
var workload_pb = require('./workload_pb.js');
var google_protobuf_struct_pb = require('google-protobuf/google/protobuf/struct_pb.js');

function serialize_JWTBundlesRequest(arg) {
    if (!(arg instanceof workload_pb.JWTBundlesRequest)) {
        throw new Error('Expected argument of type JWTBundlesRequest');
    }
    return Buffer.from(arg.serializeBinary());
}

function deserialize_JWTBundlesRequest(buffer_arg) {
    return workload_pb.JWTBundlesRequest.deserializeBinary(
        new Uint8Array(buffer_arg)
    );
}

function serialize_JWTBundlesResponse(arg) {
    if (!(arg instanceof workload_pb.JWTBundlesResponse)) {
        throw new Error('Expected argument of type JWTBundlesResponse');
    }
    return Buffer.from(arg.serializeBinary());
}

function deserialize_JWTBundlesResponse(buffer_arg) {
    return workload_pb.JWTBundlesResponse.deserializeBinary(
        new Uint8Array(buffer_arg)
    );
}

function serialize_JWTSVIDRequest(arg) {
    if (!(arg instanceof workload_pb.JWTSVIDRequest)) {
        throw new Error('Expected argument of type JWTSVIDRequest');
    }
    return Buffer.from(arg.serializeBinary());
}

function deserialize_JWTSVIDRequest(buffer_arg) {
    return workload_pb.JWTSVIDRequest.deserializeBinary(
        new Uint8Array(buffer_arg)
    );
}

function serialize_JWTSVIDResponse(arg) {
    if (!(arg instanceof workload_pb.JWTSVIDResponse)) {
        throw new Error('Expected argument of type JWTSVIDResponse');
    }
    return Buffer.from(arg.serializeBinary());
}

function deserialize_JWTSVIDResponse(buffer_arg) {
    return workload_pb.JWTSVIDResponse.deserializeBinary(
        new Uint8Array(buffer_arg)
    );
}

function serialize_ValidateJWTSVIDRequest(arg) {
    if (!(arg instanceof workload_pb.ValidateJWTSVIDRequest)) {
        throw new Error('Expected argument of type ValidateJWTSVIDRequest');
    }
    return Buffer.from(arg.serializeBinary());
}

function deserialize_ValidateJWTSVIDRequest(buffer_arg) {
    return workload_pb.ValidateJWTSVIDRequest.deserializeBinary(
        new Uint8Array(buffer_arg)
    );
}

function serialize_ValidateJWTSVIDResponse(arg) {
    if (!(arg instanceof workload_pb.ValidateJWTSVIDResponse)) {
        throw new Error('Expected argument of type ValidateJWTSVIDResponse');
    }
    return Buffer.from(arg.serializeBinary());
}

function deserialize_ValidateJWTSVIDResponse(buffer_arg) {
    return workload_pb.ValidateJWTSVIDResponse.deserializeBinary(
        new Uint8Array(buffer_arg)
    );
}

function serialize_X509SVIDRequest(arg) {
    if (!(arg instanceof workload_pb.X509SVIDRequest)) {
        throw new Error('Expected argument of type X509SVIDRequest');
    }
    return Buffer.from(arg.serializeBinary());
}

function deserialize_X509SVIDRequest(buffer_arg) {
    return workload_pb.X509SVIDRequest.deserializeBinary(
        new Uint8Array(buffer_arg)
    );
}

function serialize_X509SVIDResponse(arg) {
    if (!(arg instanceof workload_pb.X509SVIDResponse)) {
        throw new Error('Expected argument of type X509SVIDResponse');
    }
    return Buffer.from(arg.serializeBinary());
}

function deserialize_X509SVIDResponse(buffer_arg) {
    return workload_pb.X509SVIDResponse.deserializeBinary(
        new Uint8Array(buffer_arg)
    );
}

var SpiffeWorkloadAPIService = (exports.SpiffeWorkloadAPIService = {
    // JWT-SVID Profile
    fetchJWTSVID: {
        path: '/SpiffeWorkloadAPI/FetchJWTSVID',
        requestStream: false,
        responseStream: false,
        requestType: workload_pb.JWTSVIDRequest,
        responseType: workload_pb.JWTSVIDResponse,
        requestSerialize: serialize_JWTSVIDRequest,
        requestDeserialize: deserialize_JWTSVIDRequest,
        responseSerialize: serialize_JWTSVIDResponse,
        responseDeserialize: deserialize_JWTSVIDResponse,
    },
    fetchJWTBundles: {
        path: '/SpiffeWorkloadAPI/FetchJWTBundles',
        requestStream: false,
        responseStream: true,
        requestType: workload_pb.JWTBundlesRequest,
        responseType: workload_pb.JWTBundlesResponse,
        requestSerialize: serialize_JWTBundlesRequest,
        requestDeserialize: deserialize_JWTBundlesRequest,
        responseSerialize: serialize_JWTBundlesResponse,
        responseDeserialize: deserialize_JWTBundlesResponse,
    },
    validateJWTSVID: {
        path: '/SpiffeWorkloadAPI/ValidateJWTSVID',
        requestStream: false,
        responseStream: false,
        requestType: workload_pb.ValidateJWTSVIDRequest,
        responseType: workload_pb.ValidateJWTSVIDResponse,
        requestSerialize: serialize_ValidateJWTSVIDRequest,
        requestDeserialize: deserialize_ValidateJWTSVIDRequest,
        responseSerialize: serialize_ValidateJWTSVIDResponse,
        responseDeserialize: deserialize_ValidateJWTSVIDResponse,
    },
    // X.509-SVID Profile
    // Fetch all SPIFFE identities the workload is entitled to, as
    // well as related information like trust bundles and CRLs. As
    // this information changes, subsequent messages will be sent.
    fetchX509SVID: {
        path: '/SpiffeWorkloadAPI/FetchX509SVID',
        requestStream: false,
        responseStream: true,
        requestType: workload_pb.X509SVIDRequest,
        responseType: workload_pb.X509SVIDResponse,
        requestSerialize: serialize_X509SVIDRequest,
        requestDeserialize: deserialize_X509SVIDRequest,
        responseSerialize: serialize_X509SVIDResponse,
        responseDeserialize: deserialize_X509SVIDResponse,
    },
});

exports.SpiffeWorkloadAPIClient = grpc.makeGenericClientConstructor(
    SpiffeWorkloadAPIService
);
