// @universe-design/react/lib/details-display/DetailsDisplayContext.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useDetailsDisplayContext = exports.useDetailsDisplay = exports.DetailsDisplayContextProvider = void 0;
var _reactUtilities = require("@universe-design/react-utilities");
var _defaults = require("./defaults");
var _configProvider = require("../config-provider");
/**
 * detailsDisplay hook
 */
const useDetailsDisplay = props => {
  const {
    column = _defaults.detailsDisplayDefaults.column,
    layout = _defaults.detailsDisplayDefaults.layout,
    size = _defaults.detailsDisplayDefaults.size,
    colon = _defaults.detailsDisplayDefaults.colon,
    align = _defaults.detailsDisplayDefaults.align,
    data = _defaults.detailsDisplayDefaults.data,
    labelStyle = _defaults.detailsDisplayDefaults.labelStyle,
    contentStyle = _defaults.detailsDisplayDefaults.contentStyle,
    labelClassName = _defaults.detailsDisplayDefaults.labelClassName,
    contentClassName = _defaults.detailsDisplayDefaults.contentClassName
  } = props;

  // prefix
  const prefix = (0, _configProvider.useCssScope)('details-display')();

  // align
  const mergedAlign = (() => {
    if (layout === 'vertical' || layout === 'inline') {
      return 'start';
    }
    return align;
  })();
  return {
    prefix,
    column,
    layout,
    size,
    colon,
    align: mergedAlign,
    data,
    labelStyle,
    contentStyle,
    labelClassName,
    contentClassName
  };
};
exports.useDetailsDisplay = useDetailsDisplay;
/**
 * detailsDisplay 组件全局状态
 */
const {
  Provider: DetailsDisplayContextProvider,
  useContext: useDetailsDisplayContext
} = (0, _reactUtilities.createContextHook)('DetailsDisplayContext', {
  strict: true
});
exports.useDetailsDisplayContext = useDetailsDisplayContext;
exports.DetailsDisplayContextProvider = DetailsDisplayContextProvider;
