// @universe-design/react/lib/details-display/DetailsDisplayItemContext.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useDetailsDisplayItemContext = exports.useDetailsDisplayItem = exports.DetailsDisplayItemContextProvider = void 0;
var _reactUtilities = require("@universe-design/react-utilities");
var _defaults = require("./defaults");
/**
 * detailsDisplayItem hook
 */
const useDetailsDisplayItem = props => {
  const {
    label = _defaults.detailsDisplayItemDefaults.label,
    content = _defaults.detailsDisplayItemDefaults.content,
    labelStyle = _defaults.detailsDisplayItemDefaults.labelStyle,
    contentStyle = _defaults.detailsDisplayItemDefaults.contentStyle,
    labelClassName = _defaults.detailsDisplayItemDefaults.labelClassName,
    contentClassName = _defaults.detailsDisplayItemDefaults.contentClassName,
    span = _defaults.detailsDisplayItemDefaults.span,
    isRowEnd = _defaults.detailsDisplayItemDefaults.isRowEnd,
    isColumnEnd = _defaults.detailsDisplayItemDefaults.isColumnEnd
  } = props;
  return {
    label,
    content,
    labelStyle,
    contentStyle,
    labelClassName,
    contentClassName,
    span,
    isRowEnd,
    isColumnEnd
  };
};
exports.useDetailsDisplayItem = useDetailsDisplayItem;
/**
 * detailsDisplayItem 组件状态
 */
const {
  Provider: DetailsDisplayItemContextProvider,
  useContext: useDetailsDisplayItemContext
} = (0, _reactUtilities.createContextHook)('DetailsDisplayItemContext', {
  strict: true
});
exports.useDetailsDisplayItemContext = useDetailsDisplayItemContext;
exports.DetailsDisplayItemContextProvider = DetailsDisplayItemContextProvider;
