// @universe-design/react/lib/details-display/DetailsDisplay.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DetailsDisplay = void 0;
var React = _interopRequireWildcard(require("react"));
var _react2 = require("@universe-design/stylex/react");
var _DetailsDisplayContext = require("./DetailsDisplayContext");
var _sx = require("./sx");
var _Row = _interopRequireDefault(require("./Row"));
/** @jsxRuntime classic */
/** @jsx jsx */

// eslint-disable-next-line @typescript-eslint/no-unused-vars

/**
 * @docgen
 */

/**
 * @docgen
 */

const DetailsDisplay = props => {
  const {
    style,
    className,
    sx = [],
    ...useDetailsDisplayProp
  } = props;
  const contextValue = (0, _DetailsDisplayContext.useDetailsDisplay)(useDetailsDisplayProp);
  const {
    prefix,
    column,
    data
  } = contextValue;

  /**
   * format
   * 1、把每个数据的 span 格式化为 1 或者最大值 column；
   * 2、二维数组的每一个一维数组为一个tr；
   * 3、每一个数据的 span 大于 column 或者在每行的剩余空间放不下时，自动放在下一行；
   * 3.1、放到下一行后，会把上一行的最后一个数据扩大为占满剩余 tr 空间 span；
   * 4、最后输出二维数组；
   */

  const rowItems = ((originData, _column) => {
    const rows = [];
    let tmpRow = [];
    let rowRestCol = _column;
    originData.forEach((item, index) => {
      const span = Math.min(item.span || 1, _column);
      item.span = span;
      if (span <= rowRestCol) {
        tmpRow.push(item);
        rowRestCol -= span;
      } else {
        rows.push(tmpRow);
        tmpRow = [item];
        rowRestCol = _column - span;
      }
      if (rowRestCol === 0 || index === originData.length - 1) {
        rows.push(tmpRow);
        tmpRow = [];
        rowRestCol = _column;
      }
    });
    rows.forEach(item => {
      const sum = item.reduce((acc, child) => acc + (child.span || 1), 0);
      if (sum < _column) {
        item[item.length - 1].span += _column - sum;
      }
    });
    return rows;
  })(data, column);
  const styles = (0, _sx.useDetailsDisplayStylexStyles)({
    prefix
  });
  return (0, _react2.jsx)(_DetailsDisplayContext.DetailsDisplayContextProvider, {
    value: contextValue
  }, (0, _react2.jsx)("div", {
    className: className,
    sx: [styles.base, ...sx],
    style: style
  }, (0, _react2.jsx)("table", {
    sx: [styles.table]
  }, (0, _react2.jsx)("tbody", null, rowItems.map((row, r) => (0, _react2.jsx)(_Row.default, {
    key: r,
    isRowEnd: r === rowItems.length - 1,
    row: row
  }))))));
};
exports.DetailsDisplay = DetailsDisplay;
