// @universe-design/react/lib/details-display/Label.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _react2 = require("@universe-design/stylex/react");
var _clsx = _interopRequireDefault(require("clsx"));
var _hooks = require("../config-provider/hooks");
var _DetailsDisplayContext = require("./DetailsDisplayContext");
var _DetailsDisplayItemContext = require("./DetailsDisplayItemContext");
var _sx = require("./sx");
var _text = require("../text");
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars

const Label = () => {
  const locale = (0, _hooks.useLocale)();
  const {
    prefix,
    colon,
    labelStyle: rootLabelStyle,
    labelClassName: rootLabelClassName
  } = (0, _DetailsDisplayContext.useDetailsDisplayContext)();
  const {
    label,
    labelStyle,
    labelClassName
  } = (0, _DetailsDisplayItemContext.useDetailsDisplayItemContext)();
  const styles = (0, _sx.useLabelStylexStyles)({
    prefix: `${prefix}__label`
  });
  return (0, _react2.jsx)("div", {
    sx: [styles.base]
  }, (0, _react2.jsx)(_text.Text, {
    style: {
      ...rootLabelStyle,
      ...labelStyle
    },
    className: (0, _clsx.default)(rootLabelClassName, labelClassName),
    lineClamp: 2
  }, label), (0, _react2.jsx)(_text.Text, {
    sx: [styles.colon]
  }, colon && locale.global.colon));
};
var _default = exports.default = Label;
