// @universe-design/react/lib/details-display/Row.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var React = _interopRequireWildcard(require("react"));
var _react2 = require("@universe-design/stylex/react");
var _DetailsDisplayContext = require("./DetailsDisplayContext");
var _DetailsDisplayItemContext = require("./DetailsDisplayItemContext");
var _sx = require("./sx");
var _HeaderCell = _interopRequireDefault(require("./HeaderCell"));
var _DataCell = _interopRequireDefault(require("./DataCell"));
var _Label = _interopRequireDefault(require("./Label"));
var _Content = _interopRequireDefault(require("./Content"));
/* eslint-disable react/no-unknown-property */
/** @jsxRuntime classic */
/** @jsx jsx */
/** @jsxFrag React.Fragment */

// eslint-disable-next-line @typescript-eslint/no-unused-vars

const renderCells = (items, _ref) => {
  let {
    type,
    component,
    showLabel,
    showContent,
    isRowEnd
  } = _ref;
  return items.map((item, i) => {
    const {
      layout
    } = (0, _DetailsDisplayContext.useDetailsDisplayContext)();
    const layoutInline = layout === 'inline';
    // 处理cell之间的间隙，是否是最后一列
    const isColumnEnd = i === items.length - 1;
    const contextValue = (0, _DetailsDisplayItemContext.useDetailsDisplayItem)({
      ...item,
      isRowEnd,
      isColumnEnd
    });
    if (typeof component === 'string') {
      return (0, _react2.jsx)(_DetailsDisplayItemContext.DetailsDisplayItemContextProvider, {
        key: `${type}-${component}-${i}`,
        value: contextValue
      }, showLabel && !layoutInline && (0, _react2.jsx)(_HeaderCell.default, null, (0, _react2.jsx)(_Label.default, null)), showContent && (0, _react2.jsx)(_DataCell.default, null, layoutInline && (0, _react2.jsx)(_Label.default, null), (0, _react2.jsx)(_Content.default, null)));
    }
    return [(0, _react2.jsx)(_DetailsDisplayItemContext.DetailsDisplayItemContextProvider, {
      key: `${type}-${component[0]}-${i}`,
      value: {
        ...contextValue,
        span: 1
      }
    }, (0, _react2.jsx)(_HeaderCell.default, null, (0, _react2.jsx)(_Label.default, null))), (0, _react2.jsx)(_DetailsDisplayItemContext.DetailsDisplayItemContextProvider, {
      key: `${type}-${component[1]}-${i}`,
      value: {
        ...contextValue,
        span: (item.span || 1) * 2 - 1
      }
    }, (0, _react2.jsx)(_DataCell.default, null, layoutInline && (0, _react2.jsx)(_Label.default, null), (0, _react2.jsx)(_Content.default, null)))];
  });
};
const Row = props => {
  const {
    isRowEnd,
    row
  } = props;
  const {
    prefix,
    layout
  } = (0, _DetailsDisplayContext.useDetailsDisplayContext)();
  const styles = (0, _sx.useRowStylexStyles)({
    prefix: `${prefix}__row`
  });
  if (layout === 'vertical') {
    return (0, _react2.jsx)(React.Fragment, null, (0, _react2.jsx)("tr", {
      sx: [styles.base]
    }, renderCells(row, {
      type: 'label',
      component: 'th',
      showLabel: true,
      isRowEnd: false
    })), (0, _react2.jsx)("tr", {
      sx: [styles.base]
    }, renderCells(row, {
      type: 'content',
      component: 'td',
      showContent: true,
      isRowEnd
    })));
  }
  if (layout === 'inline') {
    return (0, _react2.jsx)("tr", {
      sx: [styles.base]
    }, renderCells(row, {
      type: 'inline',
      component: 'td',
      showLabel: true,
      showContent: true,
      isRowEnd
    }));
  }
  return (0, _react2.jsx)("tr", {
    sx: [styles.base]
  }, renderCells(row, {
    type: 'item',
    component: ['th', 'td'],
    isRowEnd
  }));
};
var _default = exports.default = Row;
