// @universe-design/react/lib/details-display/sx.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useRowStylexStyles = exports.useLabelStylexStyles = exports.useDetailsDisplayStylexStyles = exports.useCellStylexStyles = void 0;
var _ut0j16j = _interopRequireDefault(require("./../.stylex/ut0j16j"));
var _u5q3evl = _interopRequireDefault(require("./../.stylex/u5q3evl"));
var _ueh72r = _interopRequireDefault(require("./../.stylex/ueh72r3"));
var _ug0oh3b = _interopRequireDefault(require("./../.stylex/ug0oh3b"));
var _u1gmpn6f = _interopRequireDefault(require("./../.stylex/u1gmpn6f"));
var _u4vqhs = _interopRequireDefault(require("./../.stylex/u4vqhs6"));
var _u127hhy = _interopRequireDefault(require("./../.stylex/u127hhy3"));
var _u1ibfn = _interopRequireDefault(require("./../.stylex/u1ibfn25"));
var _u12v4awa = _interopRequireDefault(require("./../.stylex/u12v4awa"));
var _u135sb2c = _interopRequireDefault(require("./../.stylex/u135sb2c"));
var _u1co87pz = _interopRequireDefault(require("./../.stylex/u1co87pz"));
var _u1lnu8wd = _interopRequireDefault(require("./../.stylex/u1lnu8wd"));
var _u1hdaclu = _interopRequireDefault(require("./../.stylex/u1hdaclu"));
var _u22yt5q = _interopRequireDefault(require("./../.stylex/u22yt5q"));
var _u165peiw = _interopRequireDefault(require("./../.stylex/u165peiw"));
var _u24on9z = _interopRequireDefault(require("./../.stylex/u24on9z"));
var _uu1oa8u = _interopRequireDefault(require("./../.stylex/uu1oa8u"));
var _u1pwi6nc = _interopRequireDefault(require("./../.stylex/u1pwi6nc"));
var _u11kywcj = _interopRequireDefault(require("./../.stylex/u11kywcj"));
var _u1dhzo8t = _interopRequireDefault(require("./../.stylex/u1dhzo8t"));
var _u1iw9z = _interopRequireDefault(require("./../.stylex/u1iw9z84"));
var _u13d3rh = _interopRequireDefault(require("./../.stylex/u13d3rh8"));
var _u19bueuo = _interopRequireDefault(require("./../.stylex/u19bueuo"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _react = require("@universe-design/stylex/react");
//.ujw5fx7{width:16px;}
//.uq2vo04{min-width:16px;}
//.u1rvp7gu{color:@text-caption;}
//.utuwxu6{display:inline-flex;}
//.u1guyajx{align-items:baseline;}
//.unaqv0d{padding-right:40px;}
//.u15kemlc{padding-right:24px;}
//.u1y2xyjm{padding-bottom:4px;}
//.ues3tcz{text-align:right;}
//.u1o700av{text-align:left;}
//.ue2my4m{padding-bottom:20px;}
//.up67ikv{padding-bottom:16px;}
//.upe6lb7{padding-bottom:12px;}
//.u1z30l8{color:@text-title;}
//.u1nbblvp{padding-top:0px;}
//.uifp7yv{padding-right:0px;}
//.u1ov4xf1{padding-bottom:0px;}
//.u1asdtw4{padding-left:0px;}
//.uvrlu0f{min-width:44px;}
//.u82itaf{line-height:22px;}
//.u1cdqgvm{vertical-align:baseline;}
//.uly5x3f{width:100%;}
//.uses1vf{font-size:14px;}
// table
const useDetailsDisplayStylexStyles = exports.useDetailsDisplayStylexStyles = (0, _react.makeStylexStyles)(_ref => {
  let {
    prefix
  } = _ref;
  return {
    base: {
      "props": {
        "Be2twd7": _u19bueuo.default
      },
      "alias": prefix
    },
    table: {
      "props": {
        "a9b677": _u13d3rh.default
      },
      "alias": `${prefix}__table`
    }
  };
});

// row
const useRowStylexStyles = exports.useRowStylexStyles = (0, _react.makeStylexStyles)(_ref2 => {
  let {
    prefix
  } = _ref2;
  return {
    base: {
      "props": {
        "Bg96gwp": _u1dhzo8t.default,
        "ha4doy": _u1iw9z.default
      },
      "alias": prefix
    }
  };
});

// HeaderCell && DataCell
const useCellStylexStyles = exports.useCellStylexStyles = (0, _react.makeStylexStyles)(_ref3 => {
  let {
    prefix
  } = _ref3;
  return {
    /**
     * th & reset
     * minWidth：两个汉字（14 * 2） + 冒号的宽度（16）
     *  */
    th: {
      "props": {
        "z8tnut": _u165peiw.default,
        "z189sj": _u24on9z.default,
        "Byoj8tv": _uu1oa8u.default,
        "uwmqm3": _u1pwi6nc.default,
        "Bf4jedk": _u11kywcj.default
      },
      "alias": `${prefix}`
    },
    // td & reset
    td: {
      "props": {
        "z8tnut": _u165peiw.default,
        "z189sj": _u24on9z.default,
        "Byoj8tv": _uu1oa8u.default,
        "uwmqm3": _u1pwi6nc.default,
        "sj55zd": _u22yt5q.default,
        "Bf4jedk": _u11kywcj.default
      },
      "alias": `${prefix}`
    },
    // 列表三种尺寸间距
    size: {
      small: {
        "props": {
          "Byoj8tv": _u1hdaclu.default
        },
        "alias": `${prefix}-small`
      },
      medium: {
        "props": {
          "Byoj8tv": _u1lnu8wd.default
        },
        "alias": `${prefix}-medium`
      },
      large: {
        "props": {
          "Byoj8tv": _u1co87pz.default
        },
        "alias": `${prefix}-large`
      }
    },
    align: {
      left: {
        "props": {
          "fsow6f": _u135sb2c.default
        },
        "alias": `${prefix}-align-left`
      },
      right: {
        "props": {
          "fsow6f": _u12v4awa.default
        },
        "alias": `${prefix}-align-right`
      }
    },
    // 垂直布局 label 和 content 的间距
    headerGap: {
      "props": {
        "Byoj8tv": _u1ibfn.default
      },
      "alias": `${prefix}-gap`
    },
    layout: {
      // 垂直布局横向 item 间距
      verticalGap: {
        "props": {
          "z189sj": _u127hhy.default
        },
        "alias": `${prefix}-vertical-gap`
      },
      // 水平布局横向 item 间距
      horizontalGap: {
        "props": {
          "z189sj": _u4vqhs.default
        },
        "alias": `${prefix}-horizontal-gap`
      }
    },
    // layout === 'inline'时，设置display: 'inline-flex'
    inlineFlex: {
      "props": {
        "mc9l5x": _ug0oh3b.default,
        "Bt984gj": _u1gmpn6f.default
      },
      "alias": `${prefix}-inline-flex`
    }
  };
});

// label
const useLabelStylexStyles = exports.useLabelStylexStyles = (0, _react.makeStylexStyles)(_ref4 => {
  let {
    prefix
  } = _ref4;
  return {
    base: {
      "props": {
        "mc9l5x": _ug0oh3b.default,
        "Bt984gj": _u1gmpn6f.default,
        "sj55zd": _ueh72r.default
      },
      "alias": prefix
    },
    colon: {
      "props": {
        "a9b677": _ut0j16j.default,
        "Bf4jedk": _u5q3evl.default,
        "fsow6f": _u135sb2c.default
      },
      "alias": `${prefix}__colon`
    }
  };
});
