// @universe-design/react/es/cascader/CascaderOverlay.js
import React, { useContext } from 'react';
import { isNonLeaf } from '../tree';
import { SelectActionType, SelectContext } from '../select';
import useCascade from './cascade/useCascade';
import { CascadeUI } from './cascade/cascadeUI';
import { CascadeContextProvider } from './cascade/CascadeContext';
import { SelectOverlayContainer } from '../select/SelectOverlayContainer';
import clsx from 'clsx';
import { useCascaderMenuCssScope } from './common';
export const CascaderOverlay = props => {
  const cssScope = useCascaderMenuCssScope();
  const {
    inputValue,
    valueSet,
    set,
    multiple
  } = useContext(SelectContext);
  const {
    doublyTreeNodeMap,
    doublyTreeNodeList,
    changeOnSelect,
    showCheckedStrategy,
    cascadeRef,
    ...restProps
  } = props;
  const onSelect = React.useCallback((keys, _action) => {
    const action = {
      type: _action.selected ? SelectActionType.Select : SelectActionType.DeSelect,
      payload: _action.key
    };
    let closeOverlay = true;
    if (multiple) {
      closeOverlay = false;
    } else if (changeOnSelect && isNonLeaf(doublyTreeNodeMap[_action.key].__node__)) {
      closeOverlay = false;
    }
    set(keys, {
      closeOverlay,
      action
    });
  }, [changeOnSelect, doublyTreeNodeMap, multiple, set]);
  const onCheck = React.useCallback((keys, _action) => {
    const action = {
      type: _action.checked ? SelectActionType.Checked : SelectActionType.UnChecked,
      payload: _action.key
    };
    set(keys, {
      action
    });
  }, [set]);
  const cascadeCtxValue = useCascade({
    ...restProps,
    changeOnSelect,
    value: Array.from(valueSet),
    searchValue: inputValue,
    multiple,
    onSelect,
    onCheck,
    showCheckedStrategy
  }, {
    doublyTreeNodeMap,
    doublyTreeNodeList
  }, cascadeRef);
  const {
    searchFlattenTree
  } = cascadeCtxValue;
  const rootDoublyTreeNodeList = doublyTreeNodeList.filter(n => !n.parent);
  return /*#__PURE__*/React.createElement(CascadeContextProvider, {
    value: cascadeCtxValue
  }, /*#__PURE__*/React.createElement(SelectOverlayContainer, {
    className: clsx(cssScope(), (inputValue || !rootDoublyTreeNodeList.length) && cssScope('-search') // 这个样式主要是为了设置 width 为 auto
    ),
    optionsLength: rootDoublyTreeNodeList.length,
    filteredOptionsLength: searchFlattenTree.length,
    menu: /*#__PURE__*/React.createElement(CascadeUI, null)
  }));
};
