// @universe-design/react/es/cascader/common.js
import escapeRegExp from 'lodash/escapeRegExp';
import React from 'react';
import { useCssScope } from '../config-provider/hooks';
import { HighlightConsumer } from '../shared';
import { isLeaf, isNonLeaf } from '../tree';
import { getTreeNodeConfigPath } from '../tree';
import { isNonNullable, isString } from '../util';
export function useCascaderCssScope() {
  return useCssScope('cascader');
}
export function useCascaderMenuCssScope() {
  return useCssScope('cascader__menu');
}
export const defaultFilterOption = function (inputValue, node) {
  const {
    __node__: {
      title,
      value,
      label
    }
  } = node;
  // eslint-disable-next-line no-nested-ternary
  const keyword = isString(label) ? label : isString(title) ? title : value;
  return !inputValue || new RegExp(escapeRegExp(inputValue), 'gi').test(keyword) || Boolean(node.parent && defaultFilterOption(inputValue, node.parent));
};
export const getCascaderLabel = (node, options) => {
  const labelNodes = [];
  for (let c = node; c !== null; c = c.parent) {
    let labelContentNode = c.__node__.value;
    if (!isNonNullable(options.optionLabelProp)) {
      const {
        title,
        label
      } = c.__node__;
      if (isNonNullable(label)) {
        labelContentNode = label;
      } else if (isNonNullable(title)) {
        labelContentNode = title;
      }
    } else {
      if (typeof options.optionLabelProp === 'string') {
        labelContentNode = c.__node__[options.optionLabelProp];
      } else {
        labelContentNode = options.optionLabelProp(c.__node__, {
          path: getTreeNodeConfigPath(c)
        });
      }
    }
    labelNodes.unshift(/*#__PURE__*/React.createElement(React.Fragment, {
      key: c.__node__.value
    }, /*#__PURE__*/React.createElement(HighlightConsumer, {
      markRenderer: text => /*#__PURE__*/React.createElement("span", {
        className: options.highlightMarkClassName
      }, text)
    }, labelContentNode)));
    if (!options.cascaded) {
      break;
    }
    if (c.parent) {
      labelNodes.unshift(' / ');
    }
  }
  return labelNodes;
};

// 一个节点是否能被选中
export function ifCanBeSelected(_ref) {
  let {
    node,
    selectable,
    changeOnSelect,
    multiple
  } = _ref;
  /**
   * 1. 单选 select mode 父节点不可选
   * 2. 单选 select mode 父节点可选
   * 3. 多选 select mode 父节点不可选
   * 4. 多选 select mode 父节点可选（不存在）
   */
  return selectable && (isLeaf(node) || changeOnSelect) && !(multiple && isNonLeaf(node) && changeOnSelect);
}

// 默认开启虚拟滚动
export const defaultVirtual = () => true;
