// @universe-design/react/es/cascader/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
/* eslint-disable react-hooks/rules-of-hooks */

import React, { useMemo, useRef } from 'react';
import { getCascaderLabel } from './common';
import { selectGenerator } from '../select';
import { ShowCheckedStrategy, getTreeNodeConfigPath, useTreeCheck } from '../tree';
import { devWarning, genArray, isNonNullable, warning } from '../util';
import { CascaderOverlay } from './CascaderOverlay';
import { useCascaderData } from './cascade/useCascaderData';
import './style';
import { strategyMap } from './cascade/useCascade';
import { TreeCheckedProvider } from '../tree/hooks/TreeCheckedContext';
import { COMPONENT_VERSION_ID } from '../config-provider';
const RefSelect = selectGenerator({
  selectorCssScope: 'select',
  dropdownCssScope: 'cascader'
  // defaultOptionLabelProp: option => option.title || option.value,
});
const CascaderComp = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _baseSelectProps$mode, _baseSelectProps$chan;
  const {
    options = [],
    filterOption,
    filterResultSorter,
    filterResultLimit,
    filterResultRenderer,
    checkable,
    listHeight,
    dropdownLevelWidth,
    expandTrigger,
    fullPathBackfillOption = false,
    optionLabelProp: optionLabelPropProp,
    loadedKeys,
    loadData,
    onLoad,
    showCheckedStrategy = ShowCheckedStrategy.SHOW_CHILD,
    fullHeight,
    virtual,
    ...baseSelectProps
  } = props;
  const mode = (_baseSelectProps$mode = baseSelectProps.mode) !== null && _baseSelectProps$mode !== void 0 ? _baseSelectProps$mode : 'single';
  const selectRef = useRef(null);
  const cascadeRef = useRef(null);
  const {
    doublyTreeNodeList,
    doublyTreeNodeMap,
    internalOptions
  } = useCascaderData({
    options
  });
  const treeCheckedContext = useTreeCheck({
    doublyTreeNodeMap,
    defaultCheckedKeys: genArray(props.defaultValue),
    ...('value' in props ? {
      checkedKeys: genArray(props.value)
    } : {}),
    checkStrategy: strategyMap[showCheckedStrategy],
    checkStrictly: false
  }, {
    skipped: !checkable
  });
  const {
    checkedSet
  } = treeCheckedContext;
  const checkedKeys = React.useMemo(() => Object.keys(checkedSet).filter(key => checkedSet[key]), [checkedSet]);
  if (mode === 'multiple' && checkable) {
    if ('value' in props) {
      // 受控模式
      baseSelectProps.value = checkedKeys;
    } else {
      // 非受控模式
      baseSelectProps.defaultValue = checkedKeys;
    }
  }

  /**
   * Select Props
   */
  const optionLabelProp = useMemo(() => _ref => {
    let {
      value
    } = _ref;
    return doublyTreeNodeMap[value] ? getCascaderLabel(doublyTreeNodeMap[value], {
      highlightMarkClassName: '',
      cascaded: fullPathBackfillOption,
      optionLabelProp: optionLabelPropProp
    }) : value;
  }, [doublyTreeNodeMap, fullPathBackfillOption, optionLabelPropProp]);
  baseSelectProps.mode = mode;
  const changeOnSelect = (_baseSelectProps$chan = baseSelectProps.changeOnSelect) !== null && _baseSelectProps$chan !== void 0 ? _baseSelectProps$chan : mode !== 'single';
  baseSelectProps.changeOnSelect = changeOnSelect;
  process.env.NODE_ENV !== "production" ? warning(!(mode === 'single' && checkable), devWarning('Cascader', 'Single check mode does not support.')) : void 0;
  process.env.NODE_ENV !== "production" ? warning(!(mode === 'multiple' && !checkable && changeOnSelect), devWarning('Cascader', 'Multiple select mode changeOnSelect=true does not support.')) : void 0;
  process.env.NODE_ENV !== "production" ? warning(!(mode === 'multiple' && checkable && !changeOnSelect), devWarning('Cascader', 'Multiple check mode changeOnSelect=false does not support.')) : void 0;
  const overlayNode = /*#__PURE__*/React.createElement(TreeCheckedProvider, {
    value: treeCheckedContext
  }, /*#__PURE__*/React.createElement(CascaderOverlay, {
    cascadeRef: cascadeRef,
    loadData,
    loadedKeys,
    onLoad,
    dropdownLevelWidth,
    listHeight,
    filterOption,
    filterResultSorter,
    filterResultLimit,
    filterResultRenderer,
    expandTrigger,
    changeOnSelect,
    checkable,
    doublyTreeNodeList,
    doublyTreeNodeMap,
    showCheckedStrategy,
    fullHeight,
    virtual
  }));
  React.useImperativeHandle(ref, () => ({
    blur() {
      var _selectRef$current;
      (_selectRef$current = selectRef.current) === null || _selectRef$current === void 0 ? void 0 : _selectRef$current.blur();
    },
    focus() {
      var _selectRef$current2;
      (_selectRef$current2 = selectRef.current) === null || _selectRef$current2 === void 0 ? void 0 : _selectRef$current2.focus();
    },
    get input() {
      var _selectRef$current3;
      return (_selectRef$current3 = selectRef.current) === null || _selectRef$current3 === void 0 ? void 0 : _selectRef$current3.input;
    },
    scrollIntoView() {
      var _cascadeRef$current;
      for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }
      return (_cascadeRef$current = cascadeRef.current) === null || _cascadeRef$current === void 0 ? void 0 : _cascadeRef$current.scrollIntoView(...args);
    }
  }), [cascadeRef, selectRef]);
  let selectProps;
  if (baseSelectProps.mode === 'multiple') {
    selectProps = {
      ...baseSelectProps,
      mode: 'multiple',
      options: internalOptions,
      overlayNode,
      onChange: (value, meta) => {
        if (baseSelectProps.onChange) {
          const path = value.length ? value.reduce((a, b, index) => {
            const _path = getTreeNodeConfigPath(doublyTreeNodeMap[b]);
            a[index] = _path;
            return a;
          }, []) : [];
          baseSelectProps.onChange(value, {
            ...meta,
            path
          });
        }
      }
    };
  } else {
    selectProps = {
      ...baseSelectProps,
      mode: 'single',
      options: internalOptions,
      overlayNode,
      onChange: (value, meta) => {
        if (baseSelectProps.onChange) {
          const path = getTreeNodeConfigPath(isNonNullable(value) ? doublyTreeNodeMap[value] : null);
          baseSelectProps.onChange(value, {
            ...meta,
            path
          });
        }
      }
    };
  }
  return /*#__PURE__*/React.createElement(RefSelect, _extends({}, selectProps, {
    optionLabelProp: optionLabelProp,
    ref: selectRef
  }));
});
export const Cascader = Object.assign(CascaderComp, {
  SHOW_PARENT: ShowCheckedStrategy.SHOW_PARENT,
  SHOW_CHILD: ShowCheckedStrategy.SHOW_CHILD,
  [COMPONENT_VERSION_ID]: "3.28.3"
});
if (process.env.NODE_ENV !== "production") {
  Cascader.displayName = 'Cascader';
}
