// @universe-design/styles/cjs/core/index.light.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.coretoken_video_component_drag_bar_bg_normal = exports.coretoken_settings_shadow_keyboard_down_normal = exports.coretoken_settings_notice_banner_bg_general = exports.coretoken_settings_color_tag_text = exports.coretoken_settings_color_tag_bg_up = exports.coretoken_settings_color_tag_bg_down = exports.coretoken_settings_bg_keyboard_down_normal = exports.coretoken_newtab_color_app_bg = exports.coretoken_navigation_icon_bg_hover = exports.coretoken_navigation_btn_bg_normal = exports.coretoken_information_bg_bell = exports.coretoken_groupshare_bg_placeholder = exports.coretoken_card_float = exports.coretoken_all_watermark_general = void 0;
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
exports.coretoken_all_watermark_general = 'rgba(143, 149, 158, 0.12)';
exports.coretoken_card_float = '#ffffff';
exports.coretoken_groupshare_bg_placeholder = 'linear-gradient(90deg, #4c88ff 0%, #2e65d1 100%)';
exports.coretoken_information_bg_bell = '#ffffff';
exports.coretoken_navigation_btn_bg_normal = '#ffffff';
exports.coretoken_navigation_icon_bg_hover = '#f5f6f7';
exports.coretoken_newtab_color_app_bg = '#ffffff';
exports.coretoken_settings_bg_keyboard_down_normal = '#f2f3f5';
exports.coretoken_settings_color_tag_bg_down = '#fad355';
exports.coretoken_settings_color_tag_bg_up = '#fcdf7e';
exports.coretoken_settings_color_tag_text = '#6f4a01';
exports.coretoken_settings_notice_banner_bg_general = '#f2f3f5';
exports.coretoken_settings_shadow_keyboard_down_normal = '#dee0e3';
exports.coretoken_video_component_drag_bar_bg_normal = '#1f2329';
