// @universe-design/react/lib/pagination/internal/Options.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _u1glo1sh = _interopRequireDefault(require("./../../.stylex/u1glo1sh"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireDefault(require("react"));
var _KeyCode = _interopRequireDefault(require("./KeyCode"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
//.u1pbl6dw{font-variant-numeric:tabular-nums;}
class Options extends _react.default.Component {
  constructor() {
    super(...arguments);
    (0, _defineProperty2.default)(this, "state", {
      goInputText: ''
    });
    (0, _defineProperty2.default)(this, "buildOptionText", value => `${value} ${this.props.locale.items_per_page}`);
    (0, _defineProperty2.default)(this, "changeSize", value => {
      this.props.changeSize(Number(value));
    });
    (0, _defineProperty2.default)(this, "handleChange", e => {
      this.setState({
        goInputText: e.target.value
      });
    });
    (0, _defineProperty2.default)(this, "handleBlur", e => {
      const {
        goButton,
        quickGo,
        rootPrefixCls
      } = this.props;
      const {
        goInputText
      } = this.state;
      if (goButton || goInputText === '') {
        return;
      }
      this.setState({
        goInputText: ''
      });
      if (e.relatedTarget && (e.relatedTarget.className.indexOf(`${rootPrefixCls}-item-link`) >= 0 || e.relatedTarget.className.indexOf(`${rootPrefixCls}-item`) >= 0)) {
        return;
      }
      quickGo === null || quickGo === void 0 ? void 0 : quickGo(this.getValidValue());
    });
    (0, _defineProperty2.default)(this, "go", e => {
      const {
        goInputText
      } = this.state;
      if (goInputText === '') {
        return;
      }
      if ('keyCode' in e && e.keyCode === _KeyCode.default.ENTER || e.type === 'click') {
        var _this$props$quickGo, _this$props;
        this.setState({
          goInputText: ''
        });
        (_this$props$quickGo = (_this$props = this.props).quickGo) === null || _this$props$quickGo === void 0 ? void 0 : _this$props$quickGo.call(_this$props, this.getValidValue());
      }
    });
  }
  getValidValue() {
    const {
      goInputText
    } = this.state;
    return !goInputText || isNaN(Number(goInputText)) ? undefined : Number(goInputText);
  }
  getPageSizeOptions() {
    const {
      pageSize,
      pageSizeOptions
    } = this.props;
    if (pageSizeOptions.some(option => option.toString() === pageSize.toString())) {
      return pageSizeOptions;
    }
    return pageSizeOptions.concat([pageSize.toString()]).sort((a, b) => {
      const numberA = isNaN(Number(a)) ? 0 : Number(a);
      const numberB = isNaN(Number(b)) ? 0 : Number(b);
      return numberA - numberB;
    });
  }
  render() {
    const {
      pageSize,
      locale,
      rootPrefixCls,
      changeSize,
      quickGo,
      goButton,
      selectComponentClass,
      selectPrefixCls,
      disabled
    } = this.props;
    const {
      goInputText
    } = this.state;
    const prefixCls = `${rootPrefixCls}-options`;
    const Select = selectComponentClass;
    let changeSelect = null;
    let goInput = null;
    let gotoButton = null;
    if (!changeSize && !quickGo) {
      return null;
    }
    const pageSizeOptions = this.getPageSizeOptions();
    if (changeSize && Select) {
      const options = pageSizeOptions.map((opt, i) => /*#__PURE__*/_react.default.createElement(Select.Option, {
        key: i,
        value: opt.toString(),
        className: (0, _stylex.default)({
          "props": {
            "B7salou": _u1glo1sh.default
          }
        })
      }, this.buildOptionText(opt)));
      changeSelect = /*#__PURE__*/_react.default.createElement(Select, {
        disabled: disabled,
        prefixCls: selectPrefixCls,
        showSearch: false,
        className: `${prefixCls}-size-changer`,
        optionLabelProp: "children",
        dropdownMatchSelectWidth: false,
        value: (pageSize || pageSizeOptions[0]).toString(),
        onChange: this.changeSize,
        getPopupContainer: triggerNode => triggerNode.parentNode
      }, options);
    }
    if (quickGo) {
      if (goButton) {
        gotoButton = typeof goButton === 'boolean' ? /*#__PURE__*/_react.default.createElement("button", {
          type: "button",
          onClick: this.go,
          onKeyUp: this.go,
          disabled: disabled,
          className: `${prefixCls}-quick-jumper-button`
        }, locale.jump_to_confirm) : /*#__PURE__*/_react.default.createElement("span", {
          onClick: this.go,
          onKeyUp: this.go
        }, goButton);
      }
      goInput = /*#__PURE__*/_react.default.createElement("div", {
        className: `${prefixCls}-quick-jumper`
      }, locale.jump_to, /*#__PURE__*/_react.default.createElement("input", {
        disabled: disabled,
        type: "text",
        value: goInputText,
        onChange: this.handleChange,
        onKeyUp: this.go,
        onBlur: this.handleBlur
      }), locale.page, gotoButton);
    }
    return /*#__PURE__*/_react.default.createElement("li", {
      className: prefixCls
    }, changeSelect, goInput);
  }
}
var _default = exports.default = Options;
