// @universe-design/react/lib/pagination/internal/Pager.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
const Pager = props => {
  const {
    className
  } = props;
  const prefixCls = `${props.rootPrefixCls}-item`;
  const cls = (0, _clsx.default)(prefixCls, `${prefixCls}-${props.page}`, {
    [`${prefixCls}-active`]: props.active,
    [`${prefixCls}-disabled`]: !props.page
  }, className);
  return /*#__PURE__*/_react.default.createElement("li", {
    title: props.showTitle ? String(props.page) : undefined,
    className: cls,
    onClick: () => {
      props.onClick(props.page);
    },
    onKeyPress: e => {
      var _props$onKeyPress;
      (_props$onKeyPress = props.onKeyPress) === null || _props$onKeyPress === void 0 ? void 0 : _props$onKeyPress.call(props, e, props.onClick, props.page);
    },
    tabIndex: 0
  }, props.itemRender(props.page, 'page', /*#__PURE__*/_react.default.createElement("a", {
    rel: "nofollow"
  }, props.page)));
};
var _default = exports.default = Pager;
