// @universe-design/react/lib/input/components/Counter.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = Counter;
var _clsx = _interopRequireDefault(require("clsx"));
var _react = _interopRequireDefault(require("react"));
var _configProvider = require("../../config-provider");
var _tag = require("../../tag");
function Counter(props) {
  const {
    showCount,
    count,
    maxLength,
    disabled,
    tagSize = 'extra-small'
  } = props;
  const cssScope = (0, _configProvider.useCssScope)('input');
  if (!showCount) {
    return null;
  }
  return /*#__PURE__*/_react.default.createElement(_tag.Tag, {
    size: tagSize,
    className: cssScope('__count'),
    color: disabled ? 'neutral-option' : 'neutral-solid',
    disabled: disabled,
    style: props.style
  }, /*#__PURE__*/_react.default.createElement("span", {
    className: (0, _clsx.default)(count && maxLength && maxLength < count && cssScope('__count-exceed'), count && cssScope('__count-current'))
  }, count), maxLength && /*#__PURE__*/_react.default.createElement("span", null, "/", maxLength));
}
