// @universe-design/react/src/form/stories/homeless.stories.tsx
/** @jsxImportSource @emotion/react */
// also exported from '@storybook/react' if you can deal with breaking changes in 6.1
import React, { useRef } from 'react';
import type { Meta } from '@storybook/react';
import {
  Form,
  Input,
  Button,
  InputNumber,
  Select,
  Slider,
  Rate,
  Switch,
  Radio,
  Checkbox,
  FormList,
  useFormList,
  FormItem,
  useFormContextSelector,
  Table,
  Text,
} from '@universe-design/react';
import AddOutlined from '@universe-design/icons-react/AddOutlined';
import type { FormContextType } from '../../react-mini-form';
import { FormSelectors } from '../../react-mini-form';
import { FormItemUI } from '../FormItemUI';

export default {
  title: 'Components/Form/临时示例',
  component: Form,
} satisfies Meta;

const textButtonNode = (
  <Button icon={<AddOutlined />} color="primary" type="text">
    文字按钮
  </Button>
);

export const UxReview = () => (
  <Form<any> css={{ width: 600 }} initialValues={{}}>
    <Form.ItemUI label="标题" required tooltip="辅助信息">
      <Input placeholder="请输入文字" />
    </Form.ItemUI>
    <Form.ItemUI label="标题" required tooltip="辅助信息" extra={textButtonNode}>
      <Input placeholder="请输入文字" />
    </Form.ItemUI>
    <Form.ItemUI label="标题" required tooltip="辅助信息">
      <InputNumber
        css={{
          width: '100%',
        }}
        placeholder="请输入数字"
      />
    </Form.ItemUI>
    <Form.ItemUI label="标题" required tooltip="辅助信息">
      <Input addonBefore="Http://" placeholder="请输入文字" />
    </Form.ItemUI>
    <Form.ItemUI label="标题" required tooltip="辅助信息" validateStatus="error" help="输入错误">
      <Input error value="输入错误" />
    </Form.ItemUI>
    <Form.ItemUI
      label="标题"
      required
      tooltip="辅助信息"
      validateStatus="error"
      help="报错提示文案"
      extra={textButtonNode}
    >
      <Input placeholder="请输入文字" />
    </Form.ItemUI>
    <Form.ItemUI label="标题" required tooltip="辅助信息" validateStatus="error" help="输入类型错误">
      <InputNumber css={{ width: '100%' }} error value={'A' as unknown as number} />
    </Form.ItemUI>
    <Form.ItemUI label="标题" required tooltip="辅助信息" validateStatus="error" help="输入错误">
      <Input addonBefore="Http://" error value="输入错误" />
    </Form.ItemUI>
    <Form.ItemUI label="标题" required tooltip="辅助信息" validateStatus="error" help="输入错误">
      <Input.TextArea error value="报错报错报错报错报错报错报错报错报错报错" />
    </Form.ItemUI>
    <Form.ItemUI label="标题" required tooltip="辅助信息" validateStatus="error" help="输入错误">
      <Select error value="错误文案" />
    </Form.ItemUI>
    <Form.ItemUI label="标题" required tooltip="辅助信息" validateStatus="error" help="输入错误">
      <Slider value={150 / 400} max={1} />
    </Form.ItemUI>
    <Form.ItemUI label="标题" required tooltip="辅助信息" validateStatus="error" help="输入错误">
      <Rate showDescriptions={false} />
    </Form.ItemUI>
    <Form.ItemUI label="标题" required tooltip="辅助信息" validateStatus="error" help="输入错误">
      <Switch defaultChecked={true} />
    </Form.ItemUI>
    <Form.ItemUI label="标题" required tooltip="辅助信息" validateStatus="error" help="输入错误">
      <Radio.Group
        // error
        options={[
          {
            value: 0,
            label: '未选中项',
          },
          {
            value: 1,
            label: '未选中项',
          },
          {
            value: 2,
            label: '未选中项',
          },
        ]}
      />
    </Form.ItemUI>
    <Form.ItemUI label="标题" required tooltip="辅助信息" validateStatus="error" help="输入错误">
      <Checkbox.Group
        error
        options={[
          {
            value: 0,
            label: '未选中项',
          },
          {
            value: 1,
            label: '未选中项',
          },
          {
            value: 2,
            label: '未选中项',
          },
        ]}
      />
    </Form.ItemUI>
  </Form>
);

export function MutipleRulesBug() {
  return (
    <Form layout="horizontal" initialValues={{}}>
      <Form.Item
        label="多个规则校验"
        name="test"
        rules={[
          {
            required: true,
          },
          {
            max: 5,
          },
        ]}
      >
        <Input />
      </Form.Item>
    </Form>
  );
}

export function EEZhanShixiaBug() {
  const form = useRef<FormContextType<any>>(null);

  const onFinish = (values: any) => {
    console.log('Success:', values);
  };

  return (
    <div className="container">
      <Form layout="horizontal" initialValues={{ a: '' }} onFinish={onFinish} ref={form}>
        <FormSelectors
          selectors={{
            values: (ctx: FormContextType<any>) => ctx.values,
          }}
        >
          {({ values }) => (
            <Form.Item
              label="test"
              name={['test', 'test1']}
              // validateTrigger={['onSubmit', 'onChange']}
              rules={[
                {
                  validator: (_, valueInfo, callback) => {
                    console.log('test-1', valueInfo);
                    if (valueInfo !== '1') {
                      callback();
                      return true;
                    }
                    callback('error');
                    return false;
                  },
                },
              ]}
            >
              <Input style={{ height: 32, width: '100%' }} />
            </Form.Item>
          )}
        </FormSelectors>

        <Form.Item>
          <Button type="primary" htmlType="submit">
            Submit
          </Button>
        </Form.Item>
      </Form>
    </div>
  );
}

export function FormWithTextarea() {
  return (
    <Form layout="horizontal" initialValues={{}}>
      <Form.Item label="TextArea">
        <Input.TextArea />
      </Form.Item>
      <Form.Item label="Input">
        <Input />
      </Form.Item>
      <Form.Item label="Input">
        <Input />
      </Form.Item>
      <Form.Item label="Input">
        <Input />
      </Form.Item>
    </Form>
  );
}

export function Issue471() {
  return (
    <Form initialValues={{}}>
      <Form.Item
        name="code"
        label="商品编码"
        rules={{
          max: 2,
        }}
      >
        <Input />
      </Form.Item>
    </Form>
  );
}

export function ValuesChange() {
  const form = useRef<
    FormContextType<{
      router?: string;
      method?: string;
      multi?: Array<number>;
    }>
  >(null);

  const handleValueChange = (changedValues: { router?: string; method?: string; multi?: Array<number> }) => {
    if ('router' in changedValues) {
      // 测试三个setValue
      // form.current?.setValues({ method: undefined });
      form.current?.setFieldsValue({ method: undefined, multi: [2] });
      // form.current?.setFieldValue('method', undefined);
    }
  };

  return (
    <div className="App" style={{ width: 600, margin: '0 auto', paddingTop: 32 }}>
      <Form ref={form} initialValues={{}} layout="horizontal" onValuesChange={handleValueChange}>
        <Form.Item label="router" name="router">
          <Select
            options={[
              { label: 'router1', value: '1' },
              { label: 'router2', value: '2' },
            ]}
          />
        </Form.Item>
        <Form.Item label="method" name="method">
          <Select
            options={[
              { label: 'GET', value: 'GET' },
              { label: 'POST', value: 'POST' },
            ]}
          />
        </Form.Item>
        <Form.Item label="multi" name="multi">
          <Checkbox.Group
            options={[
              {
                value: 0,
                label: '0',
              },
              {
                value: 1,
                label: '1',
              },
              {
                value: 2,
                label: '2',
              },
            ]}
          />
        </Form.Item>
      </Form>
    </div>
  );
}

export const LiYiHuiDemo = () => {
  const formRef = React.useRef<FormContextType<{ a: string }>>(null);
  return (
    <Form ref={formRef} initialValues={{ a: '1' }}>
      <Form.Item validateTrigger="xxx" name="a">
        <Input onBlur={() => formRef.current?.setFieldError('a', 'Error')} />
      </Form.Item>
    </Form>
  );
};

class DataItem {
  uid: string;
  name?: string;
  value?: string;
  constructor(name?: string, value?: string) {
    this.name = name;
    this.value = value;
    this.uid = Math.random().toString(36).substr(2, 5);
  }
}

export const FormListDemo = () => {
  const initData = {
    // 也可以是这种数据结构：['name1', 'name2']
    // 但是这种数据结构就没办法使用 uid 来复用行，某一些情况下可能有问题
    list: [new DataItem('defaultName', 'defaultValue')],
  };
  const formRef = useRef<any>();

  const rules = [
    {
      validator(_rule: any, value: any, callback: any) {
        if (value) {
          return callback();
        }
        return callback(new Error('请填写'));
      },
    },
  ];

  return (
    <Form
      ref={formRef}
      initialValues={initData}
      onValuesChange={values => {
        console.log('[onValuesChange]', values);
      }}
    >
      <FormList
        name="list"
        rowKey="uid"
        rules={[
          {
            validator(_rule, value, callback) {
              if (value?.length) {
                return callback();
              }
              return callback(new Error('至少填写一行'));
            },
          },
        ]}
        validate={value => (value?.length ? '' : '至少输入一行')}
      >
        {({ fields, error, actions }) => (
          <>
            <FormItemUI required={true} label="列表" help={error} validateStatus="error" style={{ padding: 16 }}>
              <>
                {fields.map((field, index) => (
                  <div key={field.key} style={{ display: 'grid', gridAutoFlow: 'column', gap: 8 }}>
                    <FormItem label={`name-${index}`} name={field.name.concat('name')} rules={rules}>
                      <Input />
                    </FormItem>
                    <FormItem label={`value-${index}`} name={field.name.concat('value')} rules={rules}>
                      <Input />
                    </FormItem>
                    <div style={{ paddingBlockStart: 30 }}>
                      <Button type="text" color="danger" onClick={() => field.remove()}>
                        删除
                      </Button>
                      <Button
                        type="text"
                        color="primary"
                        onClick={() =>
                          formRef.current?.validateFields([field.name.concat('name'), field.name.concat('value')])
                        }
                      >
                        校验
                      </Button>
                    </div>
                  </div>
                ))}
                <Button type="text" color="primary" onClick={() => actions.push(new DataItem('新增的行'))}>
                  新增一行
                </Button>
                <Button type="text" color="primary" onClick={() => actions.unshift(new DataItem('unshift'))}>
                  在前面插入一行
                </Button>
                <Button type="text" color="primary" onClick={() => actions.swap(0, 1)}>
                  交换第一和第二行
                </Button>
                <Button type="text" color="primary" onClick={() => formRef.current?.validateFields()}>
                  校验
                </Button>
              </>
            </FormItemUI>
          </>
        )}
      </FormList>
    </Form>
  );
};

const UseFormListDemoInner = () => {
  const { error, actions, fields } = useFormList({
    name: 'list',
    rowKey: 'uid',
    validate: [
      {
        validate(value) {
          return value?.length ? '' : '至少输入一行';
        },
      },
    ],
  });

  const validateFields = useFormContextSelector(ctx => ctx.validateFields);

  const rules = [
    {
      validator(_rule: any, value: any, callback: any) {
        if (value) {
          return callback();
        }
        return callback(new Error('请填写'));
      },
    },
  ];

  return (
    <FormItemUI required={true} label="列表" help={error} validateStatus="error" style={{ padding: 16 }}>
      <>
        {fields.map((field, index) => (
          <div key={field.key} style={{ display: 'grid', gridAutoFlow: 'column', gap: 8 }}>
            <FormItem label={`name-${index}`} name={field.name.concat('name')} rules={rules}>
              <Input />
            </FormItem>
            <FormItem label={`value-${index}`} name={field.name.concat('value')} rules={rules}>
              <Input />
            </FormItem>
            <div style={{ paddingBlockStart: 30 }}>
              <Button type="text" color="danger" onClick={() => field.remove()}>
                删除
              </Button>
              <Button
                type="text"
                color="primary"
                onClick={() => validateFields([field.name.concat('name'), field.name.concat('value')])}
              >
                校验
              </Button>
            </div>
          </div>
        ))}
        <Button type="text" color="primary" onClick={() => actions.push(new DataItem('新增的行'))}>
          新增一行
        </Button>
        <Button type="text" color="primary" onClick={() => actions.unshift(new DataItem('unshift'))}>
          在前面插入一行
        </Button>
        <Button type="text" color="primary" onClick={() => actions.swap(0, 1)}>
          交换第一和第二行
        </Button>
        <Button type="text" color="primary" onClick={() => validateFields()}>
          校验
        </Button>
      </>
    </FormItemUI>
  );
};

export const UseFormListDemo = () => {
  const initData = {
    // 也可以是这种数据结构：['name1', 'name2']
    // 但是这种数据结构就没办法使用 uid 来复用行，某一些情况下可能有问题
    list: [new DataItem('defaultName', 'defaultValue')],
  };

  return (
    <Form initialValues={initData}>
      <UseFormListDemoInner />
    </Form>
  );
};

const UseFormListTableDemoInner = () => {
  const { fields, error, actions } = useFormList({
    name: 'list',
    rowKey: 'uid',
    validate: [
      {
        validate(value) {
          return value?.length ? '' : '至少输入一行';
        },
      },
    ],
  });

  const validateFields = useFormContextSelector(ctx => ctx.validateFields);

  const rules = [
    {
      validator(_rule: any, value: any, callback: any) {
        if (value) {
          return callback();
        }
        return callback(new Error('请填写'));
      },
    },
  ];

  return (
    <Table
      rowKey="key"
      dataSource={fields}
      columns={[
        {
          title: 'name',
          render(_val, field) {
            return (
              <FormItem rules={rules} name={field.name.concat('name')}>
                <Input />
              </FormItem>
            );
          },
        },
        {
          title: 'value',
          render(_val, field) {
            return (
              <FormItem rules={rules} name={field.name.concat('value')}>
                <Input />
              </FormItem>
            );
          },
        },
        {
          title: 'actions',
          render(_val, field) {
            return (
              <div style={{ marginBlockEnd: 28 }}>
                <Button color="danger" type="text" onClick={() => field.remove()}>
                  Delete
                </Button>
              </div>
            );
          },
        },
      ]}
      footer={() => (
        <div>
          <div style={{ display: 'grid', gap: 16, gridAutoFlow: 'column' }}>
            <Button color="primary" onClick={() => actions.push(new DataItem('pushItem'))}>
              Add
            </Button>
            <Button color="primary" onClick={() => actions.unshift(new DataItem('unshiftItem'))}>
              Unshift
            </Button>
            <Button color="primary" onClick={() => actions.swap(1, 2)}>
              交换2/3行
            </Button>
            <Button color="primary" onClick={() => validateFields()}>
              Validate
            </Button>
          </div>
          <FormItemUI help={error} validateStatus="error">
            <></>
          </FormItemUI>
        </div>
      )}
      pagination={false}
    />
  );
};

export const UseFormListTableDemo = () => {
  const initData = {
    // 也可以是这种数据结构：['name1', 'name2']
    // 但是这种数据结构就没办法使用 uid 来复用行，某一些情况下可能有问题
    list: [new DataItem('defaultName', 'defaultValue')],
  };

  return (
    <Form initialValues={initData}>
      <UseFormListTableDemoInner />
    </Form>
  );
};

export const AllKindsOfValidators = () => (
  <div
    style={{
      maxWidth: 600,
    }}
  >
    <Text>这个例子用来测试几种校验：正则、枚举、长度限制、必填</Text>
    <Form
      layout="horizontal"
      initialValues={{
        username: '',
      }}
    >
      <Form.Item
        required
        label="Username"
        name="username"
        rules={[
          {
            required: true,
            message: 'Please input your username!',
          },
          {
            pattern: /^[a-z]+$/,
          },
          {
            type: 'enum',
            enum: ['a', 'aa', 'abc'],
          },
          {
            max: 3,
            min: 2,
          },
        ]}
      >
        <Input />
      </Form.Item>
    </Form>
  </div>
);

export const FormItemGlobalReRender = () => {
  const form = useRef<FormContextType<any>>(null);
  return (
    <>
      <div>
        这个例子测试: FormItemUI 使用 useContextSelector 减少重渲的问题修复。可以通过 React Developer Tools 验证。
      </div>
      <Form ref={form} initialValues={{}}>
        {Array(200)
          .fill(0)
          .map((item, index) => (
            <FormItem key={index} label={index} required name={index}>
              <Input placeholder={`placeholder1${index}`} />
            </FormItem>
          ))}
      </Form>
    </>
  );
};
