// @universe-design/react/src/form/stories/screenshot.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C4_3DynamicFormItemsTableWithFormList from '../demo/4.3.dynamic-form-items-table-with-form-list';
import CAsync from '../demo/async';
import CErrorDisplay from '../demo/error-display';
import CExtraScreenshot from '../demo/extra.screenshot';
import CFormSelectors from '../demo/form-selectors';
import CInlineVerticalAlign from '../demo/inline-vertical-align';
import CLayoutHorizontalLeftScreenshot from '../demo/layout-horizontal-left.screenshot';
import CLayoutHorizontalRightScreenshot from '../demo/layout-horizontal-right.screenshot';
import CLayoutInlineScreenshot from '../demo/layout-inline.screenshot';
import CLayoutVerticalScreenshot from '../demo/layout-vertical.screenshot';
import CSetErrors from '../demo/set-errors';
import CUseField from '../demo/use-field';
if (!(C4_3DynamicFormItemsTableWithFormList as Story).storyName) {
  (C4_3DynamicFormItemsTableWithFormList as Story).storyName = '4 3 dynamic form items table with form list';
}
if (!(CAsync as Story).storyName) {
  (CAsync as Story).storyName = 'Async';
}
if (!(CErrorDisplay as Story).storyName) {
  (CErrorDisplay as Story).storyName = 'Error display';
}
if (!(CExtraScreenshot as Story).storyName) {
  (CExtraScreenshot as Story).storyName = 'Extra screenshot';
}
if (!(CFormSelectors as Story).storyName) {
  (CFormSelectors as Story).storyName = 'Form selectors';
}
if (!(CInlineVerticalAlign as Story).storyName) {
  (CInlineVerticalAlign as Story).storyName = 'Inline vertical align';
}
if (!(CLayoutHorizontalLeftScreenshot as Story).storyName) {
  (CLayoutHorizontalLeftScreenshot as Story).storyName = 'Layout horizontal left screenshot';
}
if (!(CLayoutHorizontalRightScreenshot as Story).storyName) {
  (CLayoutHorizontalRightScreenshot as Story).storyName = 'Layout horizontal right screenshot';
}
if (!(CLayoutInlineScreenshot as Story).storyName) {
  (CLayoutInlineScreenshot as Story).storyName = 'Layout inline screenshot';
}
if (!(CLayoutVerticalScreenshot as Story).storyName) {
  (CLayoutVerticalScreenshot as Story).storyName = 'Layout vertical screenshot';
}
if (!(CSetErrors as Story).storyName) {
  (CSetErrors as Story).storyName = 'Set errors';
}
if (!(CUseField as Story).storyName) {
  (CUseField as Story).storyName = 'Use field';
}

export default {
  title: 'Components/Form/截图测试',
  tags: ['autodocs'],
};

export {
  C4_3DynamicFormItemsTableWithFormList,
  CAsync,
  CErrorDisplay,
  CExtraScreenshot,
  CFormSelectors,
  CInlineVerticalAlign,
  CLayoutHorizontalLeftScreenshot,
  CLayoutHorizontalRightScreenshot,
  CLayoutInlineScreenshot,
  CLayoutVerticalScreenshot,
  CSetErrors,
  CUseField,
};
