// @universe-design/react/src/form/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Basic from '../demo/0.basic';
import C1FormRef from '../demo/1.formRef';
import C10InlineLogin from '../demo/10.inline-login';
import C11NormalLogin from '../demo/11.normal-login';
import C12Register from '../demo/12.register';
import C14FormInModal from '../demo/14.form-in-modal';
import C16TimeRelatedControls from '../demo/16.time-related-controls';
import C17ValidateTrigger from '../demo/17.validate-trigger';
import C18ErrorDisplay from '../demo/18.error-display';
import C19UseFormAndFormProvider from '../demo/19.use-form-and-form-provider';
import C3_1RequiredMark from '../demo/3.1.required-mark';
import C3_2Size from '../demo/3.2.size';
import C3FormLayout from '../demo/3.formLayout';
import C4_2DynamicFormItemsComplex from '../demo/4.2.dynamic-form-items-complex';
import C4_3DynamicFormItemsWithFormList from '../demo/4.3.dynamic-form-items-with-form-list';
import C5NestMessages from '../demo/5.nest-messages';
import C6_1CustomizedFormControls from '../demo/6.1.customized-form-controls';
import C6ComplexFormControl from '../demo/6.complex-form-control';
if (!(C0Basic as Story).storyName) {
  (C0Basic as Story).storyName = '0. 基础用法';
  (C0Basic as Story).parameters = {
    docs: {
      description: {
        story: '简单的用法。基本的表单数据域控制展示，包含布局、初始化、验证、提交。\n\n',
      },
    },
  };
}
if (!(C1FormRef as Story).storyName) {
  (C1FormRef as Story).storyName = '1. 表单方法调用';
  (C1FormRef as Story).parameters = {
    docs: {
      description: {
        story: '在`form`外部，可以通过`ref`对表单数据域进行交互。\n\n',
      },
    },
  };
}
if (!(C10InlineLogin as Story).storyName) {
  (C10InlineLogin as Story).storyName = '10. 内联登录栏';
  (C10InlineLogin as Story).parameters = {
    docs: {
      description: {
        story: '内联登录栏，常用在顶部导航栏中。\n\n',
      },
    },
  };
}
if (!(C11NormalLogin as Story).storyName) {
  (C11NormalLogin as Story).storyName = '11. 登录框';
  (C11NormalLogin as Story).parameters = {
    docs: {
      description: {
        story: '普通的登录框，可以容纳更多的元素。\n\n',
      },
    },
  };
}
if (!(C12Register as Story).storyName) {
  (C12Register as Story).storyName = '12. 注册新用户';
  (C12Register as Story).parameters = {
    docs: {
      description: {
        story: '用户填写必须的信息以注册新用户。\n\n',
      },
    },
  };
}
if (!(C14FormInModal as Story).storyName) {
  (C14FormInModal as Story).storyName = '14. 弹出层中的新建表单';
  (C14FormInModal as Story).parameters = {
    docs: {
      description: {
        story:
          '当用户访问一个展示了某个列表的页面，想新建一项但又不想跳转页面时，可以用对话框弹出一个表单，用户填写必要信息后创建新的项。\n\n',
      },
    },
  };
}
if (!(C16TimeRelatedControls as Story).storyName) {
  (C16TimeRelatedControls as Story).storyName = '16. 时间类控件';
  (C16TimeRelatedControls as Story).parameters = {
    docs: {
      description: {
        story: '时间类组件的`value`类型为`moment`对象，所以在提交服务器前需要预处理。\n\n',
      },
    },
  };
}
if (!(C17ValidateTrigger as Story).storyName) {
  (C17ValidateTrigger as Story).storyName = '17. 校验的触发事件';
  (C17ValidateTrigger as Story).parameters = {
    docs: {
      description: {
        story: '用户可以指定某个表单项、`validator`、`rule`校验的触发事件\n\n',
      },
    },
  };
}
if (!(C18ErrorDisplay as Story).storyName) {
  (C18ErrorDisplay as Story).storyName = '18. 报错信息展示方式';
  (C18ErrorDisplay as Story).parameters = {
    docs: {
      description: {
        story:
          '我们内置两种报错信息展示方式「原位报错」和「挤压报错」，可以通过 `errorDisplay` 属性进行设置，默认是「原位报错」。「原位报错」不会发生布局抖动，但是初始间距会比较大，「挤压报错」会发生布局抖动，但是初始间距会比较小。\n\n',
      },
    },
  };
}
if (!(C19UseFormAndFormProvider as Story).storyName) {
  (C19UseFormAndFormProvider as Story).storyName = '19. 使用 `useForm` 和 `FormProvider`';
  (C19UseFormAndFormProvider as Story).parameters = {
    docs: {
      description: {
        story:
          '当需要在 form 子树以外的地方，根据 form values 或 errors 进行渲染时，可以使用 `useForm` 进行状态提升。\n\n',
      },
    },
  };
}
if (!(C3_1RequiredMark as Story).storyName) {
  (C3_1RequiredMark as Story).storyName = '3.1. 必选样式';
  (C3_1RequiredMark as Story).parameters = {
    docs: {
      description: {
        story: '通过`requiredMark`切换必选与可选样式。\n\n',
      },
    },
  };
}
if (!(C3_2Size as Story).storyName) {
  (C3_2Size as Story).storyName = '3.1. 表单尺寸';
  (C3_2Size as Story).parameters = {
    docs: {
      description: {
        story: '设置表单组件尺寸。\n\n',
      },
    },
  };
}
if (!(C3FormLayout as Story).storyName) {
  (C3FormLayout as Story).storyName = '2. 表单布局';
  (C3FormLayout as Story).parameters = {
    docs: {
      description: {
        story: '表单有 3 种布局。\n\n',
      },
    },
  };
}
if (!(C4_2DynamicFormItemsComplex as Story).storyName) {
  (C4_2DynamicFormItemsComplex as Story).storyName = '4.2. 复杂的动态增减表单项';
  (C4_2DynamicFormItemsComplex as Story).parameters = {
    docs: {
      description: {
        story: '这个例子演示了一个表单中包含多个表单控件的情况。\n\n',
      },
    },
  };
}
if (!(C4_3DynamicFormItemsWithFormList as Story).storyName) {
  (C4_3DynamicFormItemsWithFormList as Story).storyName = '4.3. 动态增减嵌套表单项';
  (C4_3DynamicFormItemsWithFormList as Story).parameters = {
    docs: {
      description: {
        story:
          '下面的例子演示了通过`Form.List`管理动态增减嵌套表单项的情况。\n\n\n表格形式\n\n<code src="./4.3.dynamic-form-items-table-with-form-list.tsx" />\n',
      },
    },
  };
}
if (!(C5NestMessages as Story).storyName) {
  (C5NestMessages as Story).storyName = '5. 嵌套结构与校验信息';
  (C5NestMessages as Story).parameters = {
    docs: {
      description: {
        story: '通过`name`支持嵌套数据结构。\n\n',
      },
    },
  };
}
if (!(C6_1CustomizedFormControls as Story).storyName) {
  (C6_1CustomizedFormControls as Story).storyName = '6.1. 自定义表单控件';
  (C6_1CustomizedFormControls as Story).parameters = {
    docs: {
      description: {
        story:
          '自定义或第三方的表单控件，也可以与 Form 组件一起使用。只要该组件遵循以下的约定：\n\n- 提供受控属性`value`或其它与`valuePropName`的值同名的属性；\n- 提供`onChange`事件或`trigger`的值同名的事件。\n\n',
      },
    },
  };
}
if (!(C6ComplexFormControl as Story).storyName) {
  (C6ComplexFormControl as Story).storyName = '6. 复杂一点的控件';
  (C6ComplexFormControl as Story).parameters = {
    docs: {
      description: {
        story:
          '这里演示`Form.Item`内有多个元素的使用方式。`<Form.Item name="field" />`只会对它的直接子元素绑定表单功能，例如直接包裹了 Input / Select。如果控件前后还有一些文案或样式装点，或者一个表单项内有多个控件，你可以使用内嵌的`Form.Item`完成。你可以给`Form.Item`自定义`style`进行内联布局，或者添加`noStyle`作为纯粹的无样式绑定组件。\n\n这里展示了三种典型场景：\n\n- `Username`：输入框后面有描述文案或其他组件，在`Form.Item`内使用`<Form.Item name="field" noStyle />`去绑定对应子控件；\n- `Address`：有两个控件，在`Form.Item`内使用两个`<Form.Item name="field" noStyle />`分别绑定对应控件；\n- `BirthDate`：有两个内联控件，错误信息展示各自控件下，使用两个`<Form.Item name="field" />`分别绑定对应控件，并修改`style`使其内联布局。\n\n> 注意，在 label 对应的 Form.Item 上不要在指定`name`属性，这个 Item 只作为布局作用。\n\n更复杂的封装复用方式可以参考下面的 `自定义表单控件` 演示。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Form/官网示例',
  tags: ['autodocs'],
};

export {
  C0Basic,
  C1FormRef,
  C10InlineLogin,
  C11NormalLogin,
  C12Register,
  C14FormInModal,
  C16TimeRelatedControls,
  C17ValidateTrigger,
  C18ErrorDisplay,
  C19UseFormAndFormProvider,
  C3_1RequiredMark,
  C3_2Size,
  C3FormLayout,
  C4_2DynamicFormItemsComplex,
  C4_3DynamicFormItemsWithFormList,
  C5NestMessages,
  C6_1CustomizedFormControls,
  C6ComplexFormControl,
};
