// @universe-design/react/src/illustration/index.tsx
import type { CSSProperties, ReactNode } from 'react';
import React, { isValidElement } from 'react';
import clsx from 'clsx';
import { useCssScope } from '../config-provider/hooks';
import stylex from '@universe-design/stylex';
import { cssTypographyTitle4, cssTypographyBody0 } from '../stylex';

import './style';

export interface IllustrationProps {
  /**
   * 节点样式
   */
  style?: CSSProperties;
  /**
   * 节点类名
   */
  className?: string;
  /**
   * 类名前缀
   * @defaultValue 'ud__illustration'
   */
  prefixCls?: string;
  /**
   * 设置图片样式
   */
  imageStyle?: CSSProperties;
  /**
   * 设置显示图片，为`string`时表示自定义图片链接
   */
  image: ReactNode;
  /**
   * 自定义标题内容
   */
  title?: ReactNode;
  /**
   * 自定义描述内容，设置为`false`时会隐藏描述部分
   */
  description?: ReactNode;
  children?: ReactNode;
}

export const Illustration: React.FC<IllustrationProps> = props => {
  const { style = {}, className, image, imageStyle, title, description, prefixCls, children } = props;

  const cssScope = useCssScope('illustration', prefixCls);
  const alt = typeof description === 'string' ? description : 'illustration';

  const renderImg = () => {
    // to do: 需要添加 url 正则校验，错误要打印 警告
    if (typeof image === 'string') {
      return <img style={imageStyle} src={image} alt={alt} />;
    }

    if (isValidElement<{ style?: React.CSSProperties }>(image)) {
      return React.cloneElement(image, { style: imageStyle });
    }
    return image;
  };

  const cls = clsx(cssScope(), className);

  return (
    <div style={style} className={cls}>
      <div className={cssScope('-image')}>{renderImg()}</div>
      {title && <div className={clsx(cssScope('-title'), stylex(cssTypographyTitle4))}>{title}</div>}
      {description && <div className={clsx(cssScope('-description'), stylex(cssTypographyBody0))}>{description}</div>}
      {children && <div className={cssScope('-footer')}>{children}</div>}
    </div>
  );
};
