// @universe-design/react/lib/date-picker/panels/date-body/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _moment = _interopRequireDefault(require("moment"));
var _clsx = _interopRequireDefault(require("clsx"));
var _Body = _interopRequireDefault(require("../base/Body"));
var _hooks = require("../../../config-provider/hooks");
var _cell = _interopRequireWildcard(require("../../cell"));
var _config = require("../base/config");
var _date = require("../../utils/date");
var _DateCalendarUIContext = require("../DateCalendar/DateCalendarUIContext");
var _utils = require("../DateCalendar/utils");
var _weekNumberWrapper = _interopRequireDefault(require("../../weekNumberWrapper"));
function DateBody() {
  const {
    selectedDates = [],
    selectedRange,
    pickerValue,
    previewDate,
    cellRenderer,
    startOfWeek,
    noOutOfViewWeek,
    shouldCellDisabled,
    hideWeekNumbers = true
  } = (0, _DateCalendarUIContext.useDateCalendarUIContext)();
  const prefixCls = (0, _hooks.useCssScope)('picker')();
  const locale = (0, _hooks.useLocaleReceiver)('datePicker');
  const datePanelPrefixCls = `${prefixCls}-date-panel`;
  const dateCellInteractiveAreaCls = (0, _cell.useDateCellInteractiveAreaCls)();
  const baseDate = (0, _date.getWeekStartDate)(pickerValue, startOfWeek);
  const DAYLABELS = locale.weekdaysMin;
  const dayLabels = [...DAYLABELS, ...DAYLABELS].slice(startOfWeek, startOfWeek + 7);
  let {
    rowNum
  } = _config.BodyConfig.date;
  if (noOutOfViewWeek) {
    const {
      colNum
    } = _config.BodyConfig.date;
    // 展示的天数 = 当前月天数 + 展示的前一个月天数
    const shouldShowDays = pickerValue.daysInMonth() + (pickerValue.clone().startOf('month').day() - startOfWeek + 7) % 7;
    const redundantLines = Math.floor(Math.abs(rowNum * colNum - shouldShowDays) / colNum);
    rowNum -= redundantLines;
  }
  return /*#__PURE__*/_react.default.createElement(_weekNumberWrapper.default, {
    hideWeekNumbers: hideWeekNumbers,
    rowNum: rowNum,
    colNum: _config.BodyConfig.date.colNum,
    baseDate: baseDate
  }, /*#__PURE__*/_react.default.createElement(_Body.default, (0, _extends2.default)({}, _config.BodyConfig.date, {
    rowNum: rowNum,
    autoHeight: noOutOfViewWeek,
    cellClassName: `${datePanelPrefixCls}-cell`,
    baseDate: baseDate,
    getCellStatus: date => {
      const now = (0, _moment.default)();
      // const [startDate = null, endDate = null] = rangedValue;
      const isOutOfView = !pickerValue.isSame(date, 'month');
      const isToday = date.isSame(now, 'day');
      const isRangePreview = previewDate ? previewDate(date) : false;
      const [startDate, endDate] = selectedRange || [];
      return {
        isToday,
        isDisabled: shouldCellDisabled(date, 'date'),
        isOutOfView,
        isRange: Boolean(startDate && endDate && date.isSameOrAfter(startDate, 'day') && date.isSameOrBefore(endDate, 'day')),
        isSelected: (0, _utils.isSelectCell)({
          selectedDates,
          selectedRange,
          mode: 'date',
          date
        }),
        isRangePreview
      };
    },
    getCellDate: (d, offset) => d.clone().add(offset, 'day'),
    dateRender: d => {
      const node = /*#__PURE__*/_react.default.createElement("div", {
        className: dateCellInteractiveAreaCls
      }, d.format('D'));
      if (typeof cellRenderer === 'function') {
        return cellRenderer(d, {
          viewMode: 'date',
          defaultNode: node
        });
      }
      return node;
    }
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(`${datePanelPrefixCls}-weeks`)
  }, dayLabels.map((label, index) => /*#__PURE__*/_react.default.createElement(_cell.default, {
    key: label + index,
    className: (0, _clsx.default)(`${datePanelPrefixCls}-cell`, `${datePanelPrefixCls}-week-cell`),
    mode: "text"
  }, label)))));
}
var _default = exports.default = DateBody;
