// @universe-design/react/es/tree/TreeContainer.js
import List from '@universe-design/rc-virtual-list';
import * as React from 'react';
import { useTreeContext } from './hooks/TreeContext';
import { InternalTreeNode } from './InternalTreeNode';
import { HighlightProvider } from '../shared';
import { ListItemLike } from '../shared/components/list';
export const TreeContainer = () => {
  const {
    finalTreeNodeList,
    rootCssScope,
    height,
    virtual,
    listRef,
    searchValue,
    notFoundContent
  } = useTreeContext();
  const listNode = /*#__PURE__*/React.createElement("div", {
    className: rootCssScope('__list')
  }, /*#__PURE__*/React.createElement(List, {
    itemKey: node => node.__node__.value,
    ref: listRef,
    data: finalTreeNodeList,
    height: height,
    itemHeight: 32,
    fullHeight: false
    // onScroll={onPopupScroll}
    ,
    virtual: virtual
    // onMouseEnter={onMouseEnter}
  }, node =>
  /*#__PURE__*/
  // 这里加一个空 div 是为了把内部带 margin 的元素包住，让 virtual list 能正常计算，tree，select，cascader
  React.createElement("div", null, /*#__PURE__*/React.createElement(InternalTreeNode, {
    key: node.__node__.value,
    doublyTreeNode: node
  }))));
  const notFoundNode = (() => {
    if (notFoundContent === null) {
      return null;
    }
    return /*#__PURE__*/React.createElement(ListItemLike, {
      className: rootCssScope('__empty')
    }, notFoundContent);
  })();
  return /*#__PURE__*/React.createElement(HighlightProvider, {
    value: React.useMemo(() => ({
      searchWords: searchValue ? [searchValue] : undefined,
      disabled: false
    }), [searchValue])
  }, /*#__PURE__*/React.createElement("div", {
    className: rootCssScope()
  }, finalTreeNodeList.length ? listNode : notFoundNode));
};
