// @universe-design/react/es/tree/TreeNodeLabel.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React from 'react';
import clsx from 'clsx';
import { HighlightConsumer } from '../shared';
import { useTreeContext } from './hooks/TreeContext';
import { useTreeNodeContext } from './hooks/TreeNodeContext';
import { isNonNullable } from '../util';
import { TextOverflow } from '../shared/components/text-overflow';
export const TreeNodeLabel = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    className,
    children,
    ...restProps
  } = _ref;
  const {
    rootCssScope,
    labelRenderer
  } = useTreeContext();
  const {
    doublyTreeNode: node
  } = useTreeNodeContext();
  const {
    disabled,
    label,
    value,
    title
  } = node.__node__;
  const markRenderer = React.useCallback(text => /*#__PURE__*/React.createElement("span", {
    className: clsx(!disabled && rootCssScope('__node-highlight'))
  }, text), [disabled, rootCssScope]);
  const labelNode = (() => {
    if (isNonNullable(children)) {
      return children;
    }
    if (isNonNullable(label)) {
      return label;
    }
    if (isNonNullable(title)) {
      return title;
    }
    if (labelRenderer) {
      return labelRenderer(node);
    }
    return value;
  })();
  const $labelNode = /*#__PURE__*/React.createElement(HighlightConsumer, {
    markRenderer: markRenderer
  }, labelNode);
  return /*#__PURE__*/React.createElement(TextOverflow, _extends({
    ref: ref,
    as: "div",
    className: clsx(rootCssScope('__node__label'), className),
    lineClamp: 1
  }, restProps), $labelNode);
});
