// @universe-design/react/es/tree/TreeNodeExpandButton.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React from 'react';
import clsx from 'clsx';
import { isLeafWithLoadData, useTreeNodeLoadData } from './hooks/useTreeNodeLoadData';
import { useTreeNodeContext } from './hooks/TreeNodeContext';
import { useTreeContext } from './hooks/TreeContext';
import { isFunction } from '../util';
import composeRefs from '../util/composeRefs';
import { ExpandButton } from '../shared/components/expand-button';
export const TreeNodeExpandButton = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    className,
    onClick,
    ...restProps
  } = _ref;
  const {
    rootCssScope,
    switcherIcon,
    finalTreeNodeList,
    loadData,
    loadedKeys,
    onLoad,
    setLoadedKeys,
    toggleExpanded
  } = useTreeContext();
  const {
    doublyTreeNode: node,
    expandButtonRef,
    isExpanded
  } = useTreeNodeContext();
  const mergedRef = React.useMemo(() => composeRefs(ref, expandButtonRef), [expandButtonRef, ref]);
  const {
    onInternalExpandedToggle,
    loading,
    isLeaf
  } = useTreeNodeLoadData({
    doublyTreeNode: node,
    loadData,
    loadedKeys,
    onLoad,
    setLoadedKeys,
    toggleExpanded
  });

  // 如果是第一层，并且 root 下没有 non-leaf 的节点，这时其实所有节点都是没有 expand icon 的所以可以不显示 empty-expand-icon
  const ifRootHasNoNonLeafChildren = finalTreeNodeList.filter(o => o.depth === 1).every(o => isLeafWithLoadData(o.__node__, {
    loadData,
    loadedKeys
  }));
  if (node.depth === 1 && ifRootHasNoNonLeafChildren) {
    return null;
  }
  return /*#__PURE__*/React.createElement(ExpandButton, _extends({
    ref: mergedRef,
    className: clsx(rootCssScope('__node__expandIcon'), className),
    loading: loading,
    isExpanded: isExpanded,
    onClick: e => {
      isFunction(onClick) && onClick(e);
      onInternalExpandedToggle();
    },
    icon: switcherIcon,
    iconType: "ExpandDownFilled",
    asPlaceholder: isLeaf
  }, restProps));
});
