// @universe-design/react/es/tree/util.js
/* eslint-disable max-depth */

import { devWarning, warning } from '../util';
import { Queue } from '../util/datastructure/Queue';
export function cloneMap(map) {
  const newMap = new Map();
  for (const key of map.keys()) {
    const v = map.get(key);
    v !== undefined && newMap.set(key, v);
  }
  return newMap;
}

/**
 * 遍历 tree config
 * @param tree
 * @param cb
 */
export function traverseTreeConfig(tree, cb) {
  let index = 0;
  function _traverse(_node, _parentDepth, _parent) {
    const _depth = _parentDepth + 1;
    if (!_node.children) {
      /* empty */
    }
    cb(_node, index, _depth, _parent);
    index += 1;
    for (const __node of _node.children || []) {
      _traverse(__node, _depth, _node);
    }
  }
  // 深度优先遍历
  for (const node of tree) {
    _traverse(node, 0, null);
  }
}
export function array2Set(arr) {
  const res = {};
  arr.forEach(o => res[o] = true);
  return res;
}
export function array2Map(arr, callbackfn) {
  const res = {};
  for (const item of arr) {
    const k = callbackfn(item);

    // 如果检测到重复的 key, 给出警告
    process.env.NODE_ENV !== "production" ? warning(res[k] === undefined, devWarning('tree', `Array element is repeated, ${k} already exists`)) : void 0;
    res[k] = item;
  }
  return res;
}
export function map2Array(map) {
  return Object.keys(map);
}
export function isLeaf(node) {
  return !node.children || node.children.length === 0;
}
export function isNonLeaf(node) {
  return Boolean(node.children && node.children.length !== 0);
}
export function isCheckedDisabled(node) {
  return node.disableCheckbox || node.disabled;
}

/**
 * @public
 */

export function getTreeNodeConfigPath(doublyTreeNode) {
  const res = [];
  let childDoublyTreeNode = doublyTreeNode;
  while (childDoublyTreeNode) {
    res.push(childDoublyTreeNode.__node__);
    childDoublyTreeNode = childDoublyTreeNode.parent;
  }
  return res;
}
export function isNonLeaf2(node) {
  return Boolean(node.children && node.children.length !== 0);
}

/**
 * 生成一个双向链表 map
 * @param tree
 */
export function genDoublyTreeNodeMap(tree) {
  const doublyTreeNodeList = [];
  const doublyTreeNodeWeakMap = new WeakMap();
  traverseTreeConfig(tree, (node, index, depth, parent) => {
    const wrappedNode = {
      key: node.value,
      __node__: node,
      __parent__: parent,
      depth,
      parent: null
    };
    doublyTreeNodeList.push(wrappedNode);
    doublyTreeNodeWeakMap.set(node, wrappedNode);
  });
  const doublyTreeNodeMap = array2Map(doublyTreeNodeList, o => o.__node__.value);
  doublyTreeNodeList.forEach(node => {
    var _node$__node__$childr;
    node.parent = node.__parent__ ? doublyTreeNodeWeakMap.get(node.__parent__) || null : null;
    node.children = (_node$__node__$childr = node.__node__.children) === null || _node$__node__$childr === void 0 ? void 0 : _node$__node__$childr.map(c => doublyTreeNodeWeakMap.get(c));
  });
  return {
    doublyTreeNodeMap,
    doublyTreeNodeList
  };
}
export function traverseDoublyTreeNode(node, cb) {
  for (let stack = [node];;) {
    var _current$children;
    const current = stack.pop();
    if (!current) {
      break;
    }
    cb(current);
    (_current$children = current.children) === null || _current$children === void 0 ? void 0 : _current$children.forEach(o => stack.push(o));
  }
}
export function getTreeNodesAncestors(doublyTreeNodes, filterTreeNode) {
  const nodeSet = new Set();
  doublyTreeNodes.forEach(node => {
    if (nodeSet.has(node)) {
      return;
    }
    // 如果命中则所有的祖先节点全部加入
    if (filterTreeNode(node)) {
      // 祖先节点
      for (let current = node.parent;;) {
        if (!current || nodeSet.has(current)) {
          break;
        }
        nodeSet.add(current);
        current = current.parent;
      }
    }
  });
  return nodeSet;
}

/**
 * 从所有的节点中找到筛选节点的所有父级和子级，用于展示
 * @param rootTreeNodes 所有的节点列表
 * @param filterTreeNode 筛选节点函数
 * @returns
 */
export function getTreeNodesWithAncestorsAndAllChildren(rootTreeNodes, filterTreeNode) {
  const nodeSet = new Set();
  // 使用广度优先，保证父节点先被遍历到
  const queue = new Queue(rootTreeNodes);
  for (;;) {
    var _node$children;
    const node = queue.dequeue();
    if (!node) {
      break;
    }
    // 如果命中则所有的祖先节点、子节点全部加入
    if (filterTreeNode(node)) {
      // 祖先节点和自己
      for (let current = node;;) {
        // 如果到头了就不跑了，如果在 nodeSet 中出现过说明他的父级也都已经在 nodeSet 中了，也不跑了
        if (!current || nodeSet.has(current)) {
          break;
        }
        nodeSet.add(current);
        current = current.parent;
      }
      // 子节点
      traverseDoublyTreeNode(node, child => {
        nodeSet.add(child);
      });
      continue;
    }
    (_node$children = node.children) === null || _node$children === void 0 ? void 0 : _node$children.forEach(n => {
      queue.enqueue(n);
    });
  }
  return nodeSet;
}
