// @universe-design/styles/esm/workplace/index.dark.atom.js
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
export const optoken_app_bg = '#1c1c1c';
export const optoken_app_tag_bg = '#fee7cd';
export const optoken_app_tag_text = '#db7018';
export const optoken_app_tagbot_bg = '#ffeaa3';
export const optoken_app_tagbot_text = '#8a6419';
export const optoken_app_taglink_bg = '#d9f5d6';
export const optoken_app_taglink_text = '#2ea121';
export const optoken_botprofile_bg_header = '#373737';
export const optoken_botprofile_tag_bg = 'rgba(255, 255, 255, 0.3)';
export const optoken_botprofile_tag_text = '#ffffff';
export const optoken_messagecard_bg_header_blue = 'linear-gradient(90deg, #4c88ff 0%, #3370eb 100%)';
export const optoken_messagecard_bg_header_carmine = 'linear-gradient(90deg, #db5ea4 0%, #c24a8e 100%)';
export const optoken_messagecard_bg_header_green = 'linear-gradient(90deg, #35872a 0%, #296621 100%)';
export const optoken_messagecard_bg_header_grey = 'linear-gradient(90deg, #757575 0%, #5f5f5f 100%)';
export const optoken_messagecard_bg_header_indigo = 'linear-gradient(90deg, #7b83f7 0%, #5e68e8 100%)';
export const optoken_messagecard_bg_header_orange = 'linear-gradient(90deg, #db7018 0%, #b85e1a 100%)';
export const optoken_messagecard_bg_header_purple = 'linear-gradient(90deg, #a575fa 0%, #8c55ec 100%)';
export const optoken_messagecard_bg_header_red = 'linear-gradient(90deg, #f05b56 0%, #d14642 100%)';
export const optoken_messagecard_bg_header_turquoise = 'linear-gradient(90deg, #1fa18f 0%, #198578 100%)';
export const optoken_messagecard_bg_header_violet = 'linear-gradient(90deg, #d661d6 0%, #b54ab5 100%)';
export const optoken_messagecard_bg_header_wathet = 'linear-gradient(90deg, #1099cc 0%, #1a7fab 100%)';
export const optoken_messagecard_bg_header_yellow = 'linear-gradient(90deg, #d49b0b 0%, #8a6419 100%)';
export const optoken_tab_bg_unselected = '#000000';
