// @universe-design/react/src/scroll-area/stories/homeless.stories.tsx
import type { Meta } from '@storybook/react';
import RefreshOutlined from '@universe-design/icons-react/RefreshOutlined';
import { Button, Checkbox, Divider, ScrollArea, Space, Text, Tooltip } from '@universe-design/react';
import React from 'react';

export default {
  title: 'Components/ScrollArea/临时示例',
} satisfies Meta;

const longText =
  "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum. Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum. Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum. Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum. Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum. Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.";
const horizontalSharedStyle = {
  maxHeight: 200,
  background: 'var(--bg-float)',
  color: 'var(--text-title)',
};

export const WithDivChildren = () => (
  <div style={{ width: 600 }}>
    <Divider>浏览器原生 overflow: overlay 行为</Divider>
    <div style={{ ...horizontalSharedStyle, maxHeight: 400, overflow: 'overlay' }}>
      <h1>parent</h1>
      <div style={{ ...horizontalSharedStyle, overflow: 'overlay', width: '80%' }}>{longText}</div>
    </div>
    <Divider>ScrollArea 嵌套 ScrollArea</Divider>
    <ScrollArea
      scrollY
      scrollX
      style={{ ...horizontalSharedStyle, maxHeight: 400 }}
      onMouseOver={() => {
        console.log('parent mouseover triggered');
      }}
      onMouseMove={() => {
        console.log('parent mousemove triggered');
      }}
    >
      <h1>Parent</h1>
      <h1>Parent</h1>
      <h1>Parent</h1>
      <ScrollArea
        scrollY
        scrollX
        style={{ ...horizontalSharedStyle, width: '60%' }}
        onMouseOver={() => {
          console.log('child mouseover triggered');
        }}
        onMouseMove={() => {
          console.log('child mousemove triggered');
        }}
      >
        <h1>Child</h1>

        <div>{longText}</div>
      </ScrollArea>
      <h1>Parent</h1>
      <h1>Parent</h1>
      <h1>Parent</h1>
      <h1>Parent</h1>
    </ScrollArea>
    <Divider>Bug 反馈，ScrollArea 嵌套div </Divider>
    <ScrollArea scrollY scrollX style={{ ...horizontalSharedStyle }}>
      <div>{longText}</div>
    </ScrollArea>
    <Divider>Normal</Divider>
    <ScrollArea scrollY scrollX style={{ ...horizontalSharedStyle }}>
      {longText}
    </ScrollArea>
  </div>
);

export const All = () => {
  const [showScrollArea, setShowScrollArea] = React.useState(true);
  const [alwaysShow, setAlwaysShow] = React.useState(false);
  const [scrollX, setScrollX] = React.useState(false);
  const [scrollY, setScrollY] = React.useState(true);
  const [virtual, setVirtual] = React.useState(false);

  const refresh = () => {
    setShowScrollArea(false);
    setTimeout(() => {
      setShowScrollArea(true);
    }, 200);
  };
  return (
    <>
      <Space>
        <Checkbox checked={alwaysShow} onCheckChange={checked => setAlwaysShow(checked)}>
          Always show Scrollbar
        </Checkbox>
        <Checkbox checked={virtual} onCheckChange={checked => setVirtual(checked)}>
          Virtual
        </Checkbox>
        <Checkbox checked={scrollX} onCheckChange={checked => setScrollX(checked)}>
          ScrollX
        </Checkbox>
        <Checkbox checked={scrollY} onCheckChange={checked => setScrollY(checked)}>
          ScrollY
        </Checkbox>
      </Space>

      <Divider />
      <Button type="primary" icon={<RefreshOutlined />} onClick={refresh}>
        Refresh
      </Button>
      <Divider />
      {showScrollArea && (
        <ScrollArea
          scrollX={scrollX}
          scrollY={scrollY}
          virtual={virtual}
          style={{
            ...horizontalSharedStyle,
            width: 400,
            padding: 20,
          }}
          showScrollBars={alwaysShow ? 'always' : 'whenScrolling'}
        >
          <div style={{ width: 500, background: 'orange', height: 20 }} />
          {longText}
        </ScrollArea>
      )}
    </>
  );
};

export const DynamicChildren = () => {
  const [toggled, setToggled] = React.useState(false);
  return (
    <>
      <Text as="div">
        收到过反馈，切换 children 会报错。因为 OverlayScrollbars 通过 js 操作 DOM 和 React 的更新产生冲突。
      </Text>
      <Button onClick={() => setToggled(!toggled)}>Toggle</Button>
      <ScrollArea virtual style={{ maxHeight: 300 }}>
        {toggled ? <h1>h1h11h1</h1> : <h2>h2 h2 h2</h2>}
      </ScrollArea>
    </>
  );
};

export const PopperInScrollArea = () => (
  <>
    <Text as="div">
      出现过内部嵌套 popper 组件，找不到准确的 scroll parent 的问题，可以用这个例子测试是否能滚动跟随。
    </Text>
    <ScrollArea style={{ maxHeight: 300 }} virtual>
      <div style={{ background: 'pink' }}>
        <Tooltip title="Tooltip content" defaultVisible={true}>
          <Button>Universe Design</Button>
        </Tooltip>
      </div>
    </ScrollArea>
  </>
);
