// @universe-design/react/src/scroll-area/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Basic from '../demo/0.basic';
import C1Virtual from '../demo/1.virtual';
import C2Mixin from '../demo/2.mixin';
if (!(C0Basic as Story).storyName) {
  (C0Basic as Story).storyName = '0. 基础用法';
  (C0Basic as Story).parameters = {
    docs: {
      description: {
        story: '基础用法。\n',
      },
    },
  };
}
if (!(C1Virtual as Story).storyName) {
  (C1Virtual as Story).storyName = '1. 使用虚拟滚动条';
  (C1Virtual as Story).parameters = {
    docs: {
      description: {
        story:
          '当系统滚动条不满足使用，强烈需要一个不占位滚动条时，可以通过 virtual 属性，使用虚拟滚动条。\n\n此时不推荐在 `ScrollArea` 组件上设置 padding、margin、position 等影响布局和定位的样式，防止和内部样式冲突及产生预期之外的行为。可将样式添加在子组件上或包裹一层父组件。\n',
      },
    },
  };
}
if (!(C2Mixin as Story).storyName) {
  (C2Mixin as Story).storyName = '2. 修改系统滚动条样式';
  (C2Mixin as Story).parameters = {
    docs: {
      description: {
        story:
          "我们还通过 less 的 mixin 提供了一套定义好的滚动条样式。可以通过下方的示例引入并使用这个 mixin:\n\n```less\n@import (reference) '~@universe-design/react/es/styles/mixins/index.less';\n\n.ud-scrollbar {\n  .webkit-scrollbar_mixin;\n}\n```\n\n接下来我们需要给滚动区域添加 `ud-scrollbar` 这个类名。\n\n当需要覆盖系统滚动条样式时，请意识到这些副作用：\n\n1. 在 Mac 设备上，系统滚动条默认不占位（和设置有关），修改滚动条样式后，滚动条变为始终占位；\n2. Firefox 浏览器不支持修改滚动条宽度，对宽度的自定义不会生效；\n",
      },
    },
  };
}

export default {
  title: 'Components/ScrollArea/官网示例',
  tags: ['autodocs'],
};

export { C0Basic, C1Virtual, C2Mixin };
