// @universe-design/react/lib/date-picker/panels/picker/Footer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DatePickerFooter = void 0;
var _react = _interopRequireDefault(require("react"));
var _moment = _interopRequireDefault(require("moment"));
var _Context = require("../../Context");
var _hooks = require("../../../config-provider/hooks");
var _button = require("../../../button");
var _clsx = _interopRequireDefault(require("clsx"));
const DatePickerFooter = props => {
  const {
    global: globalLocale,
    components: {
      datePicker: locale
    }
  } = (0, _hooks.useLocale)();
  const pickerFooterCssScope = (0, _hooks.useCssScope)('picker__footer');
  const datePickerCtx = (0, _Context.useDatePickerContext)();
  const {
    renderExtraFooter,
    setValue,
    showToday,
    needConfirm,
    showNow
  } = datePickerCtx;
  if (!showToday && !showNow && !needConfirm && !renderExtraFooter) {
    return null;
  }
  return /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(pickerFooterCssScope(), pickerFooterCssScope('-container'))
  }, function () {
    var _props$disabledDate;
    if (renderExtraFooter) {
      return /*#__PURE__*/_react.default.createElement("div", {
        key: "extra",
        className: pickerFooterCssScope('__extra')
      }, renderExtraFooter(datePickerCtx));
    }
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("div", {
      key: "default",
      className: (0, _clsx.default)([pickerFooterCssScope('__date'), needConfirm && pickerFooterCssScope('__date--left')])
    }, showToday && /*#__PURE__*/_react.default.createElement(_button.Button, {
      type: "text",
      color: "primary",
      disabled: (_props$disabledDate = props.disabledDate) === null || _props$disabledDate === void 0 ? void 0 : _props$disabledDate.call(props, (0, _moment.default)()),
      onClick: () => {
        setValue((0, _moment.default)(), !needConfirm);
      }
    }, locale.today), showNow && /*#__PURE__*/_react.default.createElement(_button.Button, {
      type: "text",
      color: "primary",
      onClick: () => {
        setValue((0, _moment.default)(), false);
      }
    }, locale.currentTime)), needConfirm && /*#__PURE__*/_react.default.createElement("div", {
      key: "confirm",
      className: pickerFooterCssScope('-confirm-buttons')
    }, /*#__PURE__*/_react.default.createElement(_button.Button, {
      size: "small",
      onClick: props.onCancel
    }, globalLocale.cancelText), /*#__PURE__*/_react.default.createElement(_button.Button, {
      size: "small",
      type: "primary",
      disabled: !props.value,
      onClick: props.onConfirm
    }, globalLocale.okText)));
  }());
};
exports.DatePickerFooter = DatePickerFooter;
