// @universe-design/react/lib/date-picker/panels/picker/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PickerPanel = PickerPanel;
exports.isDateValueEqual = exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _clsx = _interopRequireDefault(require("clsx"));
var _moment = _interopRequireDefault(require("moment"));
var _react = _interopRequireWildcard(require("react"));
var _noop = _interopRequireDefault(require("lodash/noop"));
var _hooks = require("../../../config-provider/hooks");
var _timePanel = _interopRequireDefault(require("../../../shared/components/picker/Comp/panels/time-panel"));
var _moment2 = _interopRequireDefault(require("../../../shared/components/picker/generate/moment"));
var _Context = require("../../Context");
var _Footer = require("./Footer");
var _ui = require("../../utils/ui");
var _date = require("../../utils/date");
var _Context2 = require("../../../shared/components/picker/Context");
var _useSyncState = _interopRequireDefault(require("../../../hooks/useSyncState"));
var _DateCalendar = require("../DateCalendar/DateCalendar");
require("../../style");
/**
 * @privateRemarks rangedValue不会对视图定位产生影响，value会
 * @public
 * @docgen
 * 日期选择面板的属性
 */

const isDateValueEqual = (a, b) => {
  if (a === b || a !== null && a !== void 0 && a.isSame(b)) {
    return true;
  }
  return false;
};
/**
 * @deprecated
 * 推荐使用 `DateCalendar` 或 `DateCalendarUI` 组件
 */
exports.isDateValueEqual = isDateValueEqual;
function PickerPanel(props) {
  const {
    className,
    style,
    value,
    mode = 'date',
    footer = true,
    defaultViewDate,
    format,
    disabledDate,
    // footerProps
    onCancel,
    onConfirm,
    // bodyProps
    onSelect,
    onDateMouseEnter: onDateMouseEnterProp,
    onViewDateChange: onPickerValueChange,
    dateRender,
    ...restBodyProps
  } = props;
  const prefixCls = (0, _hooks.useCssScope)('picker')();
  const datePanelPrefixCls = `${prefixCls}-date-panel`;
  const datePickerContext = (0, _Context.useDatePickerContext)();
  const {
    setValue,
    showTime,
    mode: localMode,
    setMode: setLocalMode
  } = datePickerContext;
  const timeOperationRef = _react.default.useRef({});
  const timePanelFormat = typeof format === 'function' ? undefined : format;
  const timePickerProps = showTime ? (0, _ui.getTimeProps)({
    ...(typeof showTime === 'object' ? showTime : {}),
    format: timePanelFormat
  }) : {};
  const defaultTimeValue = timePickerProps.defaultValue || (0, _moment.default)();

  // 当前面板切换到哪个时间点，并不是已选中的值或未提交的值
  const [pickerValue = (0, _moment.default)(), setViewDateState] = (0, _useSyncState.default)(value !== null && value !== void 0 ? value : defaultViewDate, {
    isEqual: isDateValueEqual
  });
  const setViewDate = date => {
    onPickerValueChange === null || onPickerValueChange === void 0 ? void 0 : onPickerValueChange(date, {
      mode: localMode
    });
    setViewDateState(date);
  };
  const onDateCalendarSelect = date => {
    if (!date) {
      // 不存在此情况
      return;
    }
    const newDate = (0, _date.mergeDateAndTime)(date, showTime ? value !== null && value !== void 0 ? value : defaultTimeValue : null);
    onSelect === null || onSelect === void 0 ? void 0 : onSelect(newDate);
    if (mode !== localMode) {
      setLocalMode(mode);
    }
  };
  const onSelectTime = (0, _react.useCallback)(time => {
    onSelect === null || onSelect === void 0 ? void 0 : onSelect(time);
  }, [onSelect]);
  const locale = (0, _hooks.useLocale)();
  const onDateMouseEnter = (0, _react.useCallback)(v => {
    onDateMouseEnterProp === null || onDateMouseEnterProp === void 0 ? void 0 : onDateMouseEnterProp((0, _date.mergeDateAndTime)(v, showTime ? value !== null && value !== void 0 ? value : defaultTimeValue : null));
  }, [defaultTimeValue, onDateMouseEnterProp, showTime, value]);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: className,
    style: style
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(`${datePanelPrefixCls}__wrapper`, showTime && `${datePanelPrefixCls}--withTime`)
  }, /*#__PURE__*/_react.default.createElement(_DateCalendar.DateCalendar, (0, _extends2.default)({}, restBodyProps, {
    onPickerValueChange: setViewDate,
    onDateMouseEnter: onDateMouseEnter,
    value: value || undefined,
    mode: localMode,
    pickerValue: pickerValue,
    onSelect: d => onDateCalendarSelect === null || onDateCalendarSelect === void 0 ? void 0 : onDateCalendarSelect(d),
    disabledDate: disabledDate
  }, dateRender ? {
    cellRenderer: (date, _ref) => {
      let {
        viewMode,
        defaultNode
      } = _ref;
      return dateRender(date, viewMode, defaultNode);
    }
  } : {})), showTime && /*#__PURE__*/_react.default.createElement(_Context2.PickerContext.Provider, {
    value: {
      value: value !== null && value !== void 0 ? value : undefined,
      mode: 'time',
      setMode: _noop.default,
      setValue,
      setOpened: _noop.default,
      inDatePicker: true
    }
  }, /*#__PURE__*/_react.default.createElement(_timePanel.default, (0, _extends2.default)({
    locale: locale.components.picker
  }, timePickerProps, {
    picker: "time",
    mode: "time",
    prefixCls: prefixCls,
    generateConfig: _moment2.default,
    onPanelChange: args => onSelectTime(args),
    operationRef: timeOperationRef
  })))), footer && /*#__PURE__*/_react.default.createElement(_Context.DatePickerContextProvider, {
    value: {
      ...datePickerContext,
      setValue: (date, autoCommit) => {
        setViewDate(date);
        setValue(date, autoCommit);
      }
    }
  }, /*#__PURE__*/_react.default.createElement(_Footer.DatePickerFooter, {
    value: value,
    onCancel: onCancel,
    onConfirm: onConfirm,
    disabledDate: disabledDate
  })));
}
var _default = exports.default = PickerPanel;
