// @universe-design/react/lib/date-picker/panels/DateCalendar/DateCalendarUI.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isDateValueEqual = exports.DateCalendarUI = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = require("@universe-design/stylex/react");
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _clsx = _interopRequireDefault(require("clsx"));
var _isFunction = _interopRequireDefault(require("lodash/isFunction"));
var _isNil = _interopRequireDefault(require("lodash/isNil"));
var _min = _interopRequireDefault(require("lodash/min"));
var _noop = _interopRequireDefault(require("lodash/noop"));
var _moment = _interopRequireDefault(require("moment"));
var _react2 = _interopRequireWildcard(require("react"));
var _reactUtilities = require("@universe-design/react-utilities");
var _configProvider = require("../../../config-provider");
var _space = require("../../../space");
var _forwardRef = require("../../../util/forwardRef");
var _sx = require("../../sx");
var _PanelHeaderSwitchers = require("../PanelHeaderSwitchers");
var _utils = require("../base/utils");
var _Body = _interopRequireDefault(require("../picker/Body"));
var _MonthYearSelector = require("../selector/MonthYearSelector");
var _DateCalendarUIContext = require("./DateCalendarUIContext");
require("../../style");
var _disabled = require("../../utils/disabled");
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars

const isDateValueEqual = (a, b) => {
  if ((0, _isNil.default)(a) && (0, _isNil.default)(b)) {
    return true;
  }
  if (a === b || a !== null && a !== void 0 && a.isSame(b)) {
    return true;
  }
  return false;
};
exports.isDateValueEqual = isDateValueEqual;
const defaultPreviewDate = () => false;
const defaultDisabledDate = () => false;
const DateCalendarUI = exports.DateCalendarUI = (0, _forwardRef.forwardHTMLRef)((props, ref) => {
  const {
    className,
    style,
    mode = 'date',
    pickerValue: pickerValueProp,
    hiddenCollapse,
    // monthYearSelectorProps
    onCollapsed,
    onMonthClick,
    onYearClick,
    // bodyProps
    onPickerValueChange,
    hiddenNext,
    hiddenPrev,
    headerRenderer,
    // context start
    defaultPickerValue,
    // pickerValue,
    selectedDates = [],
    selectedRange,
    onDateMouseEnter = _noop.default,
    onDateMouseLeave = _noop.default,
    onSelect = _noop.default,
    previewDate = defaultPreviewDate,
    disabledDate = defaultDisabledDate,
    minDate,
    maxDate,
    startOfWeek = _moment.default.localeData().firstDayOfWeek(),
    noOutOfViewWeek = false,
    cellRenderer,
    onContentMouseEnter = _noop.default,
    onContentMouseLeave = _noop.default,
    sx,
    janX,
    hideWeekNumbers = true,
    // context end
    ...restBodyProps
  } = props;
  const prefixCls = (0, _configProvider.useCssScope)('picker')();
  const datePanelPrefixCls = `${prefixCls}-date-panel`;
  const [localMode, setLocalMode] = _react2.default.useState(mode);
  (0, _react2.useEffect)(() => {
    if (!janX) {
      return;
    }
    const currentLocale = _moment.default.locale();
    _moment.default.updateLocale(currentLocale, {
      week: {
        dow: startOfWeek,
        doy: 7 + startOfWeek - janX
      }
    });
  }, [startOfWeek, janX]);

  // 当前面板切换到哪个时间点，并不是已选中的值或未提交的值
  const firstSelected = (0, _min.default)([...selectedDates, ...(selectedRange || [])]);
  const [pickerValue = (0, _moment.default)(), setViewDateState] = (0, _reactUtilities.useControlledState)(firstSelected !== null && firstSelected !== void 0 ? firstSelected : defaultPickerValue, pickerValueProp, {
    isEqual: isDateValueEqual
  });
  const setViewDate = date => {
    onPickerValueChange === null || onPickerValueChange === void 0 ? void 0 : onPickerValueChange(date, {
      mode: localMode
    });
    setViewDateState(date);
  };
  const shouldCellDisabled = _react2.default.useCallback((date, _mode) => (0, _disabled.shouldCellDisabled)(date, {
    minDate,
    maxDate,
    disabledDate,
    mode: _mode
  }), [disabledDate, maxDate, minDate]);
  let PrevSwitcher, NextSwitcher;
  switch (localMode) {
    case 'date':
      PrevSwitcher = _PanelHeaderSwitchers.PreMonthSwitcher;
      NextSwitcher = _PanelHeaderSwitchers.NextMonthSwitcher;
      break;
    case 'month':
      PrevSwitcher = _PanelHeaderSwitchers.PreYearSwitcher;
      NextSwitcher = _PanelHeaderSwitchers.NextYearSwitcher;
      break;
    default:
      PrevSwitcher = _PanelHeaderSwitchers.Pre20YearsSwitcher;
      NextSwitcher = _PanelHeaderSwitchers.Next20YearsSwitcher;
      break;
  }
  // #region 自定义 Header 渲染
  const monthYearSelectorNode = (0, _react.jsx)(_MonthYearSelector.MonthYearSelector, {
    mode: localMode,
    pickerValue: pickerValue,
    className: `${prefixCls}-panel-header-content__left`,
    hiddenCollapse: hiddenCollapse || localMode === 'year',
    collapsed: localMode === mode,
    onCollapsed: collapsed => {
      if ((0, _isFunction.default)(onCollapsed)) {
        onCollapsed(collapsed);
      }
      if (collapsed) {
        setLocalMode(mode);
      } else {
        const nextMode = (0, _utils.getHigherLevelMode)(mode);
        if (nextMode) {
          setLocalMode(nextMode);
        }
      }
    },
    onYearClick: () => {
      if ((0, _isFunction.default)(onYearClick)) {
        onYearClick();
      }
      setLocalMode('year');
    },
    onMonthClick: () => {
      if ((0, _isFunction.default)(onMonthClick)) {
        onMonthClick();
      }
      setLocalMode('month');
    }
  });
  const prevBtnNode = (0, _react.jsx)(PrevSwitcher, {
    onClick: () => setViewDate((0, _utils.getPrev)(pickerValue, localMode))
  });
  const nextBtnNode = (0, _react.jsx)(NextSwitcher, {
    onClick: () => setViewDate((0, _utils.getNext)(pickerValue, localMode))
  });
  const defaultNode = (0, _react.jsx)("div", {
    className: prefixCls + '-panel-header'
  }, (0, _react.jsx)("div", {
    className: `${prefixCls}-panel-header-inner`
  }, monthYearSelectorNode, (0, _react.jsx)(_space.Space, {
    size: 12,
    className: (0, _stylex.default)(_sx.sxPickerHeaderSwitchWrapper)
  }, !hiddenPrev && prevBtnNode, !hiddenNext && nextBtnNode)));
  const pickerHeaderNode = (0, _isFunction.default)(headerRenderer) ? headerRenderer({
    prevBtnNode,
    nextBtnNode,
    defaultNode
  }) : defaultNode;
  // #endregion

  return (0, _react.jsx)(_DateCalendarUIContext.CalendarUIProvider, {
    value: {
      // defaultPickerValue,
      onPickerValueChange,
      selectedDates,
      selectedRange,
      onDateMouseEnter,
      onDateMouseLeave,
      onSelect: newDate => {
        setViewDate(newDate);
        if (mode !== localMode) {
          setLocalMode(mode);
          return;
        }
        onSelect === null || onSelect === void 0 ? void 0 : onSelect(newDate);
      },
      previewDate,
      shouldCellDisabled,
      pickerValue,
      startOfWeek,
      noOutOfViewWeek,
      cellRenderer,
      onContentMouseEnter,
      onContentMouseLeave,
      janX,
      hideWeekNumbers
    }
  }, (0, _react.jsx)("div", {
    className: (0, _clsx.default)(className, datePanelPrefixCls, `${prefixCls}-date-panel__mode_${localMode}`, !hideWeekNumbers && localMode === 'date' && `${datePanelPrefixCls}__withWeekNumbers`),
    style: style,
    ref: ref,
    sx: sx
  }, pickerHeaderNode, (0, _react.jsx)(_Body.default, (0, _extends2.default)({}, restBodyProps, {
    mode: localMode
  }))));
});
