// @universe-design/react/src/pro/timezone-picker/const.tsx
import moment from '@lark/moment-timezone';
import keyBy from 'lodash/keyBy';
import once from 'lodash/once';
import type { TimeZoneRawMeta, TimeZoneMeta, TimezoneValue, ISearchEntries } from './types';
import { TIMEZONE_RAW_META_LIST } from './timezoneMeta';

export { TIMEZONE_LINK_MAP } from './timezoneLink';

const getTimezoneLocaleKey = (timezone: string) => {
  const lastSlashPos = timezone.lastIndexOf('/');
  return timezone.slice(lastSlashPos + 1);
};

export const extendRawMeta = (rawMeta: TimeZoneRawMeta): TimeZoneMeta => ({
  ...rawMeta,
  l: `Tz_${getTimezoneLocaleKey(rawMeta.v)}`,
  s: Object.entries(rawMeta.c).map(
    ([key, value]) => Object.assign([key, ...value], { c: `Ct_${key}` }) as ISearchEntries,
  ),
});

export const TIMEZONE_META_LIST = TIMEZONE_RAW_META_LIST.map(extendRawMeta);

const timezoneSorter = (a: TimeZoneRawMeta, b: TimeZoneRawMeta) => {
  const offsetA = moment().tz(a.v).utcOffset();
  const offsetB = moment().tz(b.v).utcOffset();
  return offsetA - offsetB;
};

/** 数组按照 GMT 偏移量从小到大排序 */
export const getSortedTimezoneMetaList = once(() => [...TIMEZONE_META_LIST].sort(timezoneSorter));

/** 获取 timezone meta map */
export const getTimezoneMetaMap = once(
  () => keyBy(TIMEZONE_META_LIST, item => item.v) as Record<TimezoneValue, TimeZoneMeta>,
);
