// @universe-design/react/src/pro/timezone-picker/index.tsx
import * as React from 'react';
import moment from '@lark/moment-timezone';
import isEmpty from 'lodash/isEmpty';
import { useCssScope, useLocale } from '../../config-provider/hooks';
import type { SelectOptionConfig } from '../../select';
import { Select } from '../../select';
import { Text } from '../../text';
import { useConstant } from '@universe-design/react-utilities';
import { Highlighter } from '../../shared/components/highlighter';
import { useSelectListItemCssScope } from '../../select/common';
import type { TimezonePickerProps, TimeZoneMeta } from './types';
import {
  searchTimezoneMetaList,
  fillInTimezoneMetaList,
  getTimezoneMeta,
  useTimezoneLocale,
  arrayJoin,
  useMetaTransformer,
  splitKeywords,
} from './utils';
import './style';

export const TimezonePicker: React.FC<TimezonePickerProps> = React.memo(props => {
  const defaultMetaTransformer = useMetaTransformer();
  const {
    value: valueProp,
    defaultValue: defaultValueProp,
    date,
    metaTransformer = defaultMetaTransformer,
    ...restProps
  } = props;
  const dateString = date?.toString();

  /**
   * Context
   */
  const cssScope = useSelectListItemCssScope();
  const timezonePickerCssScope = useCssScope('timezone-picker');

  /**
   * Locale
   */
  const locale = useLocale();
  const timezoneLocale = useTimezoneLocale();

  /**
   * Value
   */
  const value = React.useMemo(() => getTimezoneMeta(valueProp)?.v ?? valueProp, [valueProp]);

  // 线上逻辑，兼容时区不完整
  const defaultValue = useConstant(() => getTimezoneMeta(defaultValueProp)?.v ?? defaultValueProp);

  /**
   * Search
   */
  const [searchValue, setSearchValue] = React.useState('');
  const handleSearch = React.useCallback((inputValue: string) => {
    setSearchValue(inputValue);
  }, []);

  /**
   * Options
   */

  // 原始 meta list 的 text 和 search 字段只是占位符，需要做一下「fill in」操作
  const filledTimezoneMetaList: TimeZoneMeta[] = React.useMemo(() => {
    const _date = dateString ? moment(dateString) : undefined;
    return timezoneLocale ? fillInTimezoneMetaList(timezoneLocale, _date, metaTransformer) : [];
  }, [timezoneLocale, dateString, metaTransformer]);

  const options = React.useMemo(() => {
    // TODO 之后把 matcher 能力和 scorer 能力也放进去，label 直接基于 matches 数组去做渲染就行了
    const filteredTimezoneMetaList = searchTimezoneMetaList(filledTimezoneMetaList, searchValue);
    const searchWords = splitKeywords(searchValue);
    const highlightClassName = cssScope('__content-highlight');
    const _options: SelectOptionConfig<string>[] = filteredTimezoneMetaList.map(timezoneMeta => ({
      value: timezoneMeta.v,
      title: timezoneMeta.l,
      label: (
        <div className={timezonePickerCssScope('__label')}>
          <Highlighter
            searchWords={searchWords}
            textToHighlight={timezoneMeta.l}
            markRenderer={text => <span className={highlightClassName}>{text}</span>}
          />
          {isEmpty(timezoneMeta.cities) ? null : (
            <Text as="div" className={timezonePickerCssScope('__subtitle')} lineClamp={2}>
              <span>{timezoneLocale?.Ct_Includes}</span>
              {arrayJoin(
                timezoneMeta.cities?.map((city, index) => (
                  <span key={`item_${index}`} className={highlightClassName}>
                    {city}
                  </span>
                )),
                i => (
                  <span key={`joiner_${i}`}>、</span>
                ),
              )}
            </Text>
          )}
        </div>
      ),
    }));
    return _options;
  }, [timezoneLocale, filledTimezoneMetaList, searchValue, cssScope, timezonePickerCssScope]);

  return (
    <Select
      // 由于这里是手动控制 options，所以用 noOptionsContent 而不是 notFoundContent
      noOptionsContent={locale.global.notFoundContent}
      {...restProps}
      {...('value' in props ? { value } : {})}
      {...{
        defaultValue,
        filterOption: false,
        searchValue,
        onSearch: handleSearch,
        options,
      }}
    />
  );
});

export { chinaMainlandMetaTransformer, getTimezoneNames, getTimezoneFormatData } from './utils';
export type { TimezonePickerProps, TimezoneLabelFormatter, TimezoneFormatData, TimezoneMetaTransformer } from './types';
export type { TimezoneLabelProps } from './TimezoneLabel';
export { TimezoneLabel } from './TimezoneLabel';
