// @universe-design/react/src/pro/timezone-picker/TimezoneLabel.tsx
import * as React from 'react';
import moment from '@lark/moment-timezone';
import type { TimezoneValue, TimezoneLabelFormatter, TimezoneMetaTransformer } from './types';
import { getTimezoneFormatData, useTimezoneLocale, defaultFormatter, useMetaTransformer } from './utils';
import { getTimezoneMetaMap } from './const';

export interface TimezoneLabelProps extends React.ComponentPropsWithoutRef<'span'> {
  /**
   *  时区值
   */
  value: TimezoneValue;
  date?: moment.Moment;
  /**
   * @experimental
   */
  metaTransformer?: TimezoneMetaTransformer;
  /**
   * 自定义格式化函数
   */
  children?: TimezoneLabelFormatter;
}

export const TimezoneLabel = React.forwardRef<HTMLSpanElement, TimezoneLabelProps>((props, ref) => {
  const defaultMetaTransformer = useMetaTransformer();
  const { value, date, metaTransformer = defaultMetaTransformer, children, ...restProps } = props;
  const locale = useTimezoneLocale();
  const dateString = date?.toString();
  const formatData = React.useMemo(() => {
    if (!locale) {
      return void 0;
    }
    return getTimezoneFormatData(value, locale, {
      date: dateString ? moment(dateString) : undefined,
      cities: getTimezoneMetaMap()[value]?.c,
      transformer: metaTransformer,
    });
  }, [dateString, locale, value, metaTransformer]);
  const defaultLabel = React.useMemo(() => defaultFormatter(formatData) || value, [formatData, value]);

  return (
    <span ref={ref} {...restProps}>
      {children ? children(formatData) : defaultLabel}
    </span>
  );
});
