// @universe-design/react/src/pro/timezone-picker/timezoneMeta.tsx
// autogen by ../../../scripts/pro/gen.timezonemeta.ts
import type { TimeZoneRawMeta } from './types';

export const TIMEZONE_RAW_META_LIST: ReadonlyArray<TimeZoneRawMeta> = [
  { v: 'Pacific/Niue', c: { Niue: ['纽埃', 'niu ai'] } },
  { v: 'Pacific/Pago_Pago', c: { Samoa: ['萨摩亚', 'sa mo ya'] } },
  { v: 'Pacific/Rarotonga', c: { CookIslands: ['库克群岛', 'ku ke qun dao'] } },
  {
    v: 'Pacific/Honolulu',
    c: {
      Hawaii: ['夏威夷', 'xia wei yi'],
      Honolulu: ['檀香山', 'tan xiang shan'],
    },
  },
  { v: 'Pacific/Tahiti', c: { Tahiti: ['塔希提岛', 'ta xi ti dao'] } },
  {
    v: 'Pacific/Marquesas',
    c: { Marquesas: ['马克萨斯群岛', 'ma ke sa si qun dao'] },
  },
  {
    v: 'America/Adak',
    c: { AleutianIslands: ['阿留申群岛', 'a liu shen qun dao'] },
  },
  { v: 'Pacific/Gambier', c: { Gambier: ['甘比尔', 'gan bi er'] } },
  { v: 'America/Anchorage', c: { Anchorage: ['安克雷奇', 'an ke lei qi'] } },
  { v: 'America/Juneau', c: { Juneau: ['朱诺', 'zhu nuo'] } },
  {
    v: 'America/Metlakatla',
    c: { Metlakatla: ['梅特拉卡特拉', 'mei te la ka te la'] },
  },
  { v: 'America/Nome', c: { Nome: ['诺姆', 'nuo mu'] } },
  { v: 'America/Sitka', c: { Sitka: ['锡特卡', 'xi te ka'] } },
  { v: 'America/Yakutat', c: { Yakutat: ['亚库塔特', 'ya ku ta te'] } },
  { v: 'Pacific/Pitcairn', c: { Pitcairn: ['皮特凯恩', 'pi te kai en'] } },
  {
    v: 'America/Los_Angeles',
    c: {
      LosAngeles: ['洛杉矶', 'luo shan ji'],
      SanFrancisco: ['旧金山', 'jiu jin shan'],
      SanDiego: ['圣迭戈', 'sheng die ge'],
      Seattle: ['西雅图', 'xi ya tu'],
      LasVegas: ['拉斯维加斯', 'la si wei jia si'],
      SanJose: ['圣何塞', 'sheng he sai'],
      Portland: ['波特兰', 'bo te lan'],
      Riverside: ['河滨', 'he bin'],
      Sacramento: ['萨克拉门托', 'sa ke la men tuo'],
      MountainView: ['山景城', 'shan jing cheng'],
      Irvine: ['尔湾', 'er wan'],
    },
  },
  { v: 'America/Tijuana', c: { Tijuana: ['蒂华纳', 'di hua na'] } },
  { v: 'America/Vancouver', c: { Vancouver: ['温哥华', 'wen ge hua'] } },
  { v: 'America/Dawson', c: { Dawson: ['道森', 'dao sen'] } },
  {
    v: 'America/Whitehorse',
    c: { Whitehorse: ['怀特霍斯', 'huai te huo si'] },
  },
  { v: 'America/Hermosillo', c: { Hermosillo: ['埃莫西约', 'ai mo xi yue'] } },
  { v: 'America/Chihuahua', c: { Chihuahua: ['奇瓦瓦', 'qi wa wa'] } },
  { v: 'America/Mazatlan', c: { Mazatlan: ['马萨特兰', 'ma sa te lan'] } },
  { v: 'America/Creston', c: { Creston: ['克雷斯顿', 'ke lei si dun'] } },
  {
    v: 'America/Dawson_Creek',
    c: { DawsonCreek: ['道森克里克', 'dao sen ke li ke'] },
  },
  {
    v: 'America/Fort_Nelson',
    c: { FortNelson: ['纳尔逊堡', 'na er xun bao'] },
  },
  { v: 'America/Phoenix', c: { Phoenix: ['菲尼克斯', 'fei ni ke si'] } },
  { v: 'America/Boise', c: { Boise: ['博伊西', 'bo yi xi'] } },
  {
    v: 'America/Cambridge_Bay',
    c: { CambridgeBay: ['剑桥湾', 'jian qiao wan'] },
  },
  {
    v: 'America/Denver',
    c: { Denver: ['丹佛', 'dan fo'], SaltLakeCity: ['盐湖城', 'yan hu cheng'] },
  },
  {
    v: 'America/Edmonton',
    c: {
      Edmonton: ['埃德蒙顿', 'ai de meng dun'],
      Calgary: ['卡尔加里', 'ka er jia li'],
    },
  },
  { v: 'America/Inuvik', c: { Inuvik: ['伊努维克', 'yi nu wei ke'] } },
  { v: 'America/Ojinaga', c: { Ojinaga: ['奥希纳加', 'ao xi na jia'] } },
  {
    v: 'America/Yellowknife',
    c: { Yellowknife: ['耶洛奈夫', 'ye luo nai fu'] },
  },
  { v: 'America/Belize', c: { Belize: ['伯利兹', 'bo li zi'] } },
  {
    v: 'America/Costa_Rica',
    c: { CostaRica: ['哥斯达黎加', 'ge si da li jia'] },
  },
  {
    v: 'America/El_Salvador',
    c: {
      ElSalvador: ['萨尔瓦多', 'sa er wa duo'],
      SanSalvador: ['圣萨尔瓦多', 'sheng sa er wa duo'],
    },
  },
  {
    v: 'America/Guatemala',
    c: {
      Guatemala: ['危地马拉', 'wei di ma la'],
      GuatemalaCity: ['危地马拉城', 'wei de ma la cheng'],
    },
  },
  { v: 'America/Managua', c: { Managua: ['马那瓜', 'ma na gua'] } },
  { v: 'America/Regina', c: { Regina: ['里贾纳', 'li jia na'] } },
  {
    v: 'America/Swift_Current',
    c: { SwiftCurrent: ['斯威夫特卡伦特', 'si wei fu te ka lun te'] },
  },
  {
    v: 'America/Tegucigalpa',
    c: { Tegucigalpa: ['特古西加尔巴', 'te gu xi jia er ba'] },
  },
  {
    v: 'America/Bahia_Banderas',
    c: { BahiadeBanderas: ['巴伊亚班德拉斯', 'ba yi ya ban de la si'] },
  },
  { v: 'America/Merida', c: { Merida: ['梅里达', 'mei li da'] } },
  {
    v: 'America/Mexico_City',
    c: {
      MexicoCity: ['墨西哥城', 'mo xi ge cheng'],
      Guadalajara: ['瓜达拉哈拉', 'gua da la ha la'],
      Puebla: ['普埃布拉', 'pu ai bu la'],
    },
  },
  { v: 'America/Monterrey', c: { Monterrey: ['蒙特雷', 'meng te lei'] } },
  {
    v: 'Pacific/Galapagos',
    c: { Galapagos: ['加拉帕戈斯', 'jia la pa ge si'] },
  },
  {
    v: 'America/North_Dakota/Beulah',
    c: {
      NorthDakotaBeulah: ['北达科他州比尤拉', 'bei da ke ta zhou bi you la'],
    },
  },
  {
    v: 'America/North_Dakota/Center',
    c: {
      NorthDakotaCenterCity: ['北达科他州申特城', 'bei da ke ta zhou shen te cheng'],
    },
  },
  {
    v: 'America/Chicago',
    c: {
      Chicago: ['芝加哥', 'zhi jia ge'],
      Houston: ['休斯敦', 'xiu si dun'],
      Dallas: ['达拉斯', 'da la si'],
      KansasCity: ['堪萨斯城', 'kan sa si cheng'],
      Minneapolis: ['明尼阿波利斯', 'ming ni a bo li si'],
      Austin: ['奥斯汀', 'ao si ting'],
      NewOrleans: ['新奥尔良', 'xin ao er liang'],
      Nashville: ['纳什维尔', 'na shi wei er'],
      PortLouis: ['路易港', 'lu yi gang'],
      Birmingham: ['伯明翰', 'bo ming han'],
      SanAntonio: ['圣安东尼奥', 'sheng an dong ni ao'],
      Milwaukee: ['密尔沃基', 'mi er wo ji'],
      OklahomaCity: ['俄克拉荷马城', 'e ke la he ma cheng'],
      Memphis: ['孟菲斯', 'meng fei si'],
    },
  },
  {
    v: 'America/Indiana/Knox',
    c: { IndianaKnox: ['印第安纳州诺克斯', 'yin di an na zhou nuo ke si'] },
  },
  {
    v: 'America/Matamoros',
    c: { Matamoros: ['马塔莫罗斯', 'ma ta mo luo si'] },
  },
  { v: 'America/Menominee', c: { Menominee: ['梅诺米尼', 'mei nuo mi ni'] } },
  {
    v: 'America/North_Dakota/New_Salem',
    c: {
      NorthDakotaNewSalem: ['北达科他州新塞勒姆', 'bei da ke ta zhou xin sai le mu'],
    },
  },
  {
    v: 'America/Rainy_River',
    c: { RainyRiverDistrict: ['雷尼河', 'lei ni he'] },
  },
  { v: 'America/Rankin_Inlet', c: { RankinInlet: ['兰金因莱特', 'lan jin yin lai te'] } },
  { v: 'America/Resolute', c: { Resolute: ['雷索卢特', 'lei suo lu te'] } },
  {
    v: 'America/Indiana/Tell_City',
    c: {
      IndianaTellCity: ['印第安纳州特尔城', 'yin di an na zhou te er cheng'],
    },
  },
  { v: 'America/Winnipeg', c: { Winnipeg: ['温尼伯', 'wen ni bo'] } },
  { v: 'America/Eirunepe', c: { Eirunepe: ['埃鲁内佩', 'ai lu nei pei'] } },
  {
    v: 'America/Rio_Branco',
    c: { RioBranco: ['里奥布朗库', 'li ao bu lang ku'] },
  },
  {
    v: 'America/Bogota',
    c: { Bogota: ['波哥大', 'bo ge da'], Medellin: ['麦德林', 'mai de lin'] },
  },
  { v: 'Pacific/Easter', c: { EasterIsland: ['复活节岛', 'fu huo jie dao'] } },
  { v: 'America/Atikokan', c: { Atikokan: ['阿蒂科肯', 'a di ke ken'] } },
  { v: 'America/Cancun', c: { Cancun: ['坎昆', 'kan kun'] } },
  { v: 'America/Jamaica', c: { Kingston: ['金斯顿', 'jin si dun'] } },
  {
    v: 'America/Panama',
    c: {
      Panama: ['巴拿马', 'ba na ma'],
      PanamaCity: ['巴拿马城', 'ba na ma cheng'],
    },
  },
  {
    v: 'America/Guayaquil',
    c: {
      Ecuador: ['厄瓜多尔', 'e gua duo er'],
      Quito: ['基多', 'ji duo'],
      Guayaquil: ['瓜亚基尔', 'gua ya ji er'],
    },
  },
  { v: 'America/Lima', c: { Lima: ['利马', 'li ma'] } },
  { v: 'America/Havana', c: { Havana: ['哈瓦那', 'ha wa na'] } },
  {
    v: 'America/Detroit',
    c: {
      Detroit: ['底特律', 'di te lv'],
      GrandRapids: ['大急流城', 'da ji liu cheng'],
    },
  },
  { v: 'America/Grand_Turk', c: { GrandTurk: ['大特克', 'da te ke'] } },
  {
    v: 'America/Indiana/Indianapolis',
    c: { Indianapolis: ['印第安纳波利斯', 'yin di an na bo li si'] },
  },
  { v: 'America/Iqaluit', c: { Iqaluit: ['伊魁特', 'yi kui te'] } },
  {
    v: 'America/Kentucky/Louisville',
    c: { Louisville: ['路易斯维尔', 'lu yi si wei er'] },
  },
  {
    v: 'America/Indiana/Marengo',
    c: { IndianaMarengo: ['印第安纳州马伦戈', 'yin di an na zhou ma lun ge'] },
  },
  {
    v: 'America/Kentucky/Monticello',
    c: {
      KentuckyMonticello: ['肯塔基州蒙蒂塞洛', 'ken ta ji zhou meng di sai luo'],
    },
  },
  { v: 'America/Nassau', c: { Nassau: ['拿骚', 'na sao'] } },
  {
    v: 'America/New_York',
    c: {
      NewYork: ['纽约', 'niu yue'],
      Washington: ['华盛顿', 'hua sheng dun'],
      Boston: ['波士顿', 'bo shi dun'],
      Miami: ['迈阿密', 'mai a mi'],
      Atlanta: ['亚特兰大', 'ya te lan da'],
      Philadelphia: ['费城', 'fei cheng'],
      Baltimore: ['巴尔的摩', 'ba er di mo'],
      Orlando: ['奥兰多', 'ao lan duo'],
      Columbus: ['哥伦布', 'ge lun bu'],
      Cleveland: ['克利夫兰', 'ke li fu lan'],
      Tampa: ['坦帕', 'tan pa'],
      Cincinnati: ['辛辛那提', 'xin xin na ti'],
      Pittsburgh: ['匹兹堡', 'pi zi bao'],
      VirginiaBeach: ['弗吉尼亚比奇', 'fu ji ni ya bi qi'],
      Charlotte: ['夏洛特', 'xia luo te'],
      Buffalo: ['布法罗', 'bu fa luo'],
      Columbia: ['哥伦比亚', 'ge lun bi ya'],
      Jacksonville: ['杰克逊维尔', 'jie ke xun wei er'],
      Richmond: ['里士满', 'li shi man'],
      Rochester: ['罗彻斯特', 'luo che si te'],
      PalmBay: ['棕榈湾', 'zong lv wan'],
      Ashburn: ['阿什本', 'a shi ben'],
    },
  },
  { v: 'America/Nipigon', c: { Nipigon: ['尼皮贡', 'ni pi gong'] } },
  { v: 'America/Pangnirtung', c: { Pangnirtung: ['庞纳唐', 'pang na tang'] } },
  {
    v: 'America/Indiana/Petersburg',
    c: {
      IndianaPetersburg: ['印第安纳州彼得斯堡', 'yin di an na zhou bi de si bao'],
    },
  },
  {
    v: 'America/Port-au-Prince',
    c: { 'Port-au-Prince': ['太子港', 'tai zi gang'] },
  },
  { v: 'America/Thunder_Bay', c: { ThunderBay: ['桑德贝', 'sang de bei'] } },
  {
    v: 'America/Toronto',
    c: {
      Toronto: ['多伦多', 'duo lun duo'],
      Montreal: ['蒙特利尔', 'meng te li er'],
      Ottawa: ['渥太华', 'wo tai hua'],
      Quebec: ['魁北克', 'kui bei ke'],
      Hamilton: ['汉密尔顿', 'han mi er dun'],
    },
  },
  {
    v: 'America/Indiana/Vevay',
    c: { IndianaVevay: ['印第安纳州维维市', 'yin di an na zhou wei wei shi'] },
  },
  {
    v: 'America/Indiana/Vincennes',
    c: {
      IndianaVincennes: ['印第安纳州温森斯', 'yin di an na zhou wen sen si'],
    },
  },
  {
    v: 'America/Indiana/Winamac',
    c: {
      IndianaWinamac: ['印第安纳州威纳马克', 'yin di an na zhou wei na ma ke'],
    },
  },
  { v: 'America/Boa_Vista', c: { BoaVista: ['博阿维斯塔', 'bo a wei si ta'] } },
  { v: 'America/Manaus', c: { Manaus: ['马瑙斯', 'ma nao si'] } },
  {
    v: 'America/Porto_Velho',
    c: { PortoVelho: ['波多韦柳', 'bo duo wei liu'] },
  },
  { v: 'America/Barbados', c: { Bridgetown: ['布里奇顿', 'bu li qi dun'] } },
  {
    v: 'America/Blanc-Sablon',
    c: { 'Blanc-Sablon': ['布兰克萨布隆', 'bu lan ke sa bu long'] },
  },
  { v: 'America/Curacao', c: { Curacao: ['库拉索', 'ku la suo'] } },
  { v: 'America/Martinique', c: { Martinique: ['马提尼克', 'ma ti ni ke'] } },
  {
    v: 'America/Port_of_Spain',
    c: { PortofSpain: ['西班牙港', 'xi ban ya gang'] },
  },
  { v: 'America/Puerto_Rico', c: { PuertoRico: ['波多黎各', 'bo duo li ge'] } },
  {
    v: 'America/Santo_Domingo',
    c: { SantoDomingo: ['圣多明各', 'sheng duo ming ge'] },
  },
  { v: 'America/La_Paz', c: { Bolivia: ['玻利维亚', 'bo li wei ya'] } },
  { v: 'America/Guyana', c: { Guyana: ['圭亚那', 'gui ya na'] } },
  {
    v: 'America/Caracas',
    c: {
      Venezuela: ['委内瑞拉', 'wei nei rui la'],
      Caracas: ['加拉加斯', 'jia la jia si'],
    },
  },
  { v: 'America/Campo_Grande', c: { CampoGrande: ['大坎普', 'da kan pu'] } },
  { v: 'America/Cuiaba', c: { Cuiaba: ['库亚巴', 'ku ya ba'] } },
  { v: 'Atlantic/Bermuda', c: { Bermuda: ['百慕大', 'bai mu da'] } },
  { v: 'America/Glace_Bay', c: { GlaceBay: ['格莱斯贝', 'ge lai si bei'] } },
  { v: 'America/Goose_Bay', c: { GooseBay: ['古斯湾', 'gu si wan'] } },
  { v: 'America/Halifax', c: { Halifax: ['哈利法克斯', 'ha li fa ke si'] } },
  { v: 'America/Moncton', c: { Moncton: ['蒙克顿', 'meng ke dun'] } },
  { v: 'America/Thule', c: { Thule: ['图勒', 'tu le'] } },
  {
    v: 'America/Argentina/Buenos_Aires',
    c: {
      BuenosAires: ['布宜诺斯艾利斯', 'bu yi nuo si ai li si'],
      PortoAlegre: ['阿雷格里港', 'a lei ge li gang'],
    },
  },
  {
    v: 'America/Argentina/Catamarca',
    c: { Catamarca: ['卡塔马卡', 'ka ta ma ka'] },
  },
  {
    v: 'America/Argentina/Cordoba',
    c: { Cordoba: ['科尔多瓦', 'ke er duo wa'] },
  },
  { v: 'America/Argentina/Jujuy', c: { Jujuy: ['胡胡伊', 'hu hu yi'] } },
  {
    v: 'America/Argentina/La_Rioja',
    c: { LaRioja: ['拉里奥哈', 'la li ao ha'] },
  },
  { v: 'America/Argentina/Mendoza', c: { Mendoza: ['门多萨', 'men duo sa'] } },
  {
    v: 'America/Argentina/Rio_Gallegos',
    c: { RioGallegos: ['里奥加耶戈斯', 'li ao jia ye ge si'] },
  },
  { v: 'America/Argentina/Salta', c: { Salta: ['萨尔塔', 'sa er ta'] } },
  {
    v: 'America/Argentina/San_Juan',
    c: { ArgentinaSanJuan: ['圣胡安', 'sheng hu an'] },
  },
  { v: 'America/Argentina/Tucuman', c: { Tucuman: ['图库曼', 'tu ku man'] } },
  {
    v: 'America/Argentina/Ushuaia',
    c: { Ushuaia: ['乌斯怀亚', 'wu si huai ya'] },
  },
  {
    v: 'America/Araguaina',
    c: { Araguaina: ['阿拉瓜伊纳', 'a la gua yi na'] },
  },
  { v: 'America/Bahia', c: { Bahia: ['巴伊亚', 'ba yi ya'] } },
  { v: 'America/Belem', c: { Belem: ['贝伦', 'bei lun'] } },
  { v: 'America/Fortaleza', c: { Fortaleza: ['福塔雷萨', 'fu ta lei sa'] } },
  { v: 'America/Maceio', c: { Maceio: ['马塞约', 'ma sai yue'] } },
  { v: 'America/Recife', c: { Recife: ['累西腓', 'lei xi fei'] } },
  { v: 'America/Santarem', c: { Santarem: ['圣塔伦', 'sheng ta lun'] } },
  {
    v: 'America/Santiago',
    c: { Lima: ['利马', 'li ma'], Santiago: ['圣地亚哥', 'sheng di ya ge'] },
  },
  {
    v: 'Atlantic/Stanley',
    c: { FalklandIslands: ['福克兰群岛', 'fu ke lan qun dao'] },
  },
  {
    v: 'America/Cayenne',
    c: { FrenchGuiana: ['法属圭亚那', 'fa shu gui ya na'] },
  },
  {
    v: 'Antarctica/Palmer',
    c: { PalmerStation: ['帕默尔站', 'pa mo er zhan'] },
  },
  { v: 'America/Asuncion', c: { Paraguay: ['巴拉圭', 'ba la gui'] } },
  {
    v: 'America/Punta_Arenas',
    c: { PuntaArenas: ['蓬塔阿雷纳斯', 'peng ta a lei na si'] },
  },
  {
    v: 'Antarctica/Rothera',
    c: { RotheraStation: ['罗瑟拉站', 'luo se la zhan'] },
  },
  { v: 'America/Paramaribo', c: { Surinam: ['苏里南', 'su li nan'] } },
  {
    v: 'America/Montevideo',
    c: {
      Uruguay: ['乌拉圭', 'wu la gui'],
      Montevideo: ['蒙得维的亚', 'meng de wei de ya'],
    },
  },
  {
    v: 'America/Godthab',
    c: { WestGreenland: ['格陵兰岛西部', 'ge ling lan dao xi bu'] },
  },
  {
    v: 'America/Argentina/San_Luis',
    c: { SanLuis: ['圣路易斯', 'sheng lu yi si'] },
  },
  {
    v: 'America/Sao_Paulo',
    c: {
      RioDeJaneiro: ['里约热内卢', 'li yue re nei lu'],
      SaoPaulo: ['圣保罗', 'sheng bao luo'],
      Brasilia: ['巴西利亚', 'ba xi li ya'],
      BeloHorizonte: ['贝洛奥里藏特', 'bei luo ao li cang te'],
    },
  },
  { v: 'America/St_Johns', c: { Newfoundland: ['纽芬兰', 'niu fen lan'] } },
  {
    v: 'America/Miquelon',
    c: {
      StPierreMiquelon: ['圣皮埃尔和密克隆群岛', 'sheng pi ai er he mi ke long qun dao'],
    },
  },
  {
    v: 'America/Noronha',
    c: {
      FernandoDeNoronha: ['费尔南多-迪诺罗尼亚', 'fei er nan duo - di nuo luo ni ya'],
    },
  },
  {
    v: 'Atlantic/South_Georgia',
    c: { SouthGeorgia: ['南乔治亚岛', 'nan qiao zhi ya dao'] },
  },
  { v: 'Atlantic/Azores', c: { Azores: ['亚速尔群岛', 'ya su er qun dao'] } },
  { v: 'Atlantic/Cape_Verde', c: { CapeVerde: ['佛得角', 'fo de jiao'] } },
  {
    v: 'America/Scoresbysund',
    c: { EastGreenland: ['格陵兰岛东部', 'ge ling lan dao dong bu'] },
  },
  {
    v: 'Etc/GMT',
    c: {
      GreenwichMeanTime: ['格林尼治标准时间', 'ge lin ni zhi biao zhun shi jian'],
    },
  },
  { v: 'Africa/Abidjan', c: { Abidjan: ['阿比让', 'a bi rang'] } },
  { v: 'Africa/Accra', c: { Accra: ['阿克拉', 'a ke la'] } },
  { v: 'Africa/Bissau', c: { Bissau: ['比绍', 'bi shao'] } },
  {
    v: 'America/Danmarkshavn',
    c: { Danmarkshavn: ['丹麦港', 'dan mai gang'] },
  },
  { v: 'Africa/Monrovia', c: { Monrovia: ['蒙罗维亚', 'meng luo wei ya'] } },
  {
    v: 'Atlantic/Reykjavik',
    c: { Reykjavik: ['雷克雅未克', 'lei ke ya wei ke'] },
  },
  {
    v: 'Europe/Dublin',
    c: { Ireland: ['爱尔兰', 'ai er lan'], Dublin: ['都柏林', 'dou bo lin'] },
  },
  {
    v: 'Antarctica/Troll',
    c: { TrollStation: ['特罗尔站', 'te luo er zhan'] },
  },
  {
    v: 'Europe/London',
    c: {
      London: ['伦敦', 'lun dun'],
      Edinburgh: ['爱丁堡', 'ai ding bao'],
      Birmingham: ['伯明翰', 'bo ming han'],
      Manchester: ['曼彻斯特', 'man che si te'],
      Glasgow: ['格拉斯哥', 'ge la si ge'],
      Liverpool: ['利物浦', 'li wu pu'],
      Nottingham: ['诺丁汉', 'nuo ding han'],
      Coventry: ['考文垂', 'kao wen chui'],
      Southampton: ['南安普顿', 'nan an pu dun'],
      Belfast: ['贝尔法斯特', 'bei er fa si te'],
      MiltonKeynes: ['米尔顿凯恩斯', 'mi er dun kai en si'],
      Cardiff: ['卡迪夫', 'ka di fu'],
      Bradford: ['布拉德福德', 'bu la de fu de'],
      Leeds: ['利兹', 'li zi'],
      Newcastle: ['纽卡斯尔', 'niu ka si er'],
      Sheffield: ['谢菲尔德', 'xie fei er de'],
    },
  },
  { v: 'Atlantic/Canary', c: { Canary: ['加那利', 'jia na li'] } },
  { v: 'Atlantic/Faroe', c: { Faroe: ['法罗', 'fa luo'] } },
  {
    v: 'Europe/Lisbon',
    c: { Lisbon: ['里斯本', 'li si ben'], Porto: ['波尔图', 'bo er tu'] },
  },
  { v: 'Atlantic/Madeira', c: { Madeira: ['马德拉', 'ma de la'] } },
  { v: 'Africa/Sao_Tome', c: { SaoTome: ['圣多美', 'sheng duo mei'] } },
  {
    v: 'Africa/Algiers',
    c: {
      Algiers: ['阿尔及尔', 'a er ji er'],
      DarElBeida: ['达尔贝达', 'da er bei da'],
    },
  },
  { v: 'Africa/Tunis', c: { Tunis: ['突尼斯市', 'tu ni si shi'] } },
  {
    v: 'Europe/Amsterdam',
    c: {
      Amsterdam: ['阿姆斯特丹', 'a mu si te dan'],
      Rotterdam: ['鹿特丹', 'lu te dan'],
      TheHague: ['海牙', 'hai ya'],
      Utrecht: ['乌得勒支', 'wu de lei zhi'],
      Eindhoven: ['埃因霍温', 'ai yin huo wen'],
    },
  },
  { v: 'Europe/Andorra', c: { Andorra: ['安道尔', 'an dao er'] } },
  {
    v: 'Europe/Belgrade',
    c: {
      Belgrade: ['贝尔格莱德', 'bei er ge lai de'],
      Zagreb: ['萨格勒布', 'sa ge le bu'],
      Ljubljana: ['卢布尔雅那', 'lu bu er ya na'],
    },
  },
  {
    v: 'Europe/Berlin',
    c: {
      Berlin: ['柏林', 'bo lin'],
      Frankfurt: ['法兰克福', 'fa lan ke fu'],
      Munich: ['慕尼黑', 'mu ni hei'],
      Hamburg: ['汉堡', 'han bao'],
      Stuttgart: ['斯图加特', 'si tu jia te'],
      Dusseldorf: ['杜塞尔多夫', 'du sai er duo fu'],
      Cologne: ['科隆', 'ke long'],
      Essen: ['埃森', 'ai sen'],
      Dresden: ['德累斯顿', 'de lei si dun'],
      Hanover: ['汉诺威', 'han nuo wei'],
      Darmstadt: ['达姆施塔特', 'da mu shi ta te'],
      Karlsruhe: ['卡尔斯鲁厄', 'ka er si lu e'],
      Ingolstadt: ['因戈尔施塔特', 'yin ge er shi ta te'],
      Erlangen: ['埃朗根', 'ai lang gen'],
      Wolfsburg: ['沃尔夫斯堡', 'wo er fu si bao'],
      Regensburg: ['雷根斯堡', 'lei gen si bao'],
      Ulm: ['乌尔姆', 'wu er mu'],
      Nuremberg: ['纽伦堡', 'niu lun bao'],
      Leipzig: ['莱比锡', 'lai bi xi'],
    },
  },
  {
    v: 'Europe/Brussels',
    c: {
      Brussels: ['布鲁塞尔', 'bu lu sai er'],
      Antwerp: ['安特卫普', 'an te wei pu'],
    },
  },
  { v: 'Europe/Budapest', c: { Budapest: ['布达佩斯', 'bu da pei si'] } },
  {
    v: 'Africa/Casablanca',
    c: { Casablanca: ['卡萨布兰卡', 'ka sa bu lan ka'] },
  },
  { v: 'Africa/Ceuta', c: { Ceuta: ['休达', 'xiu da'] } },
  {
    v: 'Europe/Copenhagen',
    c: {
      Copenhagen: ['哥本哈根', 'ge ben ha gen'],
      Aarhus: ['奥胡斯', 'ao hu si'],
    },
  },
  { v: 'Africa/El_Aaiun', c: { ElAaiun: ['阿尤恩', 'a you en'] } },
  { v: 'Europe/Gibraltar', c: { Gibraltar: ['直布罗陀', 'zhi bu luo tuo'] } },
  { v: 'Europe/Luxembourg', c: { Luxembourg: ['卢森堡', 'lu sen bao'] } },
  {
    v: 'Europe/Madrid',
    c: {
      Madrid: ['马德里', 'ma de li'],
      Barcelona: ['巴塞罗那', 'ba sai luo na'],
      Valencia: ['瓦伦西亚', 'wa lun xi ya'],
      Santander: ['桑坦德', 'sang tan de'],
      Seville: ['塞维利亚', 'sai wei li ya'],
      Malaga: ['马拉加', 'ma la jia'],
      ACoruna: ['拉科鲁尼亚', 'la ke lu ni ya'],
    },
  },
  { v: 'Europe/Malta', c: { Malta: ['马耳他', 'ma er ta'] } },
  { v: 'Europe/Monaco', c: { Monaco: ['摩纳哥', 'mo na ge'] } },
  {
    v: 'Europe/Oslo',
    c: { Oslo: ['奥斯陆', 'ao si lu'], Bergen: ['卑尔根', 'bei er gen'] },
  },
  {
    v: 'Europe/Paris',
    c: {
      Paris: ['巴黎', 'ba li'],
      Lyon: ['里昂', 'li ang'],
      Marseille: ['马赛', 'ma sai'],
      Strasbourg: ['斯特拉斯堡', 'si te la si bao'],
      Lille: ['里尔', 'li er'],
      Toulouse: ['图卢兹', 'tu lu zi'],
      Nice: ['尼斯', 'ni si'],
    },
  },
  {
    v: 'Europe/Prague',
    c: {
      Prague: ['布拉格', 'bu la ge'],
      Bratislava: ['布拉迪斯拉发', 'bu la di si la fa'],
    },
  },
  {
    v: 'Europe/Rome',
    c: {
      Rome: ['罗马', 'luo ma'],
      Milan: ['米兰', 'mi lan'],
      Venice: ['威尼斯', 'wei ni si'],
      Florence: ['佛罗伦萨', 'fo luo lun sa'],
      Turin: ['都灵', 'du ling'],
      Naples: ['那不勒斯', 'na bu le si'],
      VaticanCity: ['梵蒂冈城', 'fan di gang cheng'],
      Pisa: ['比萨', 'bi sa'],
    },
  },
  {
    v: 'Europe/Stockholm',
    c: {
      Stockholm: ['斯德哥尔摩', 'si de ge er mo'],
      Malmo: ['马尔默', 'ma er mo'],
      Gothenburg: ['哥德堡', 'ge de bao'],
    },
  },
  { v: 'Europe/Tirane', c: { Tirane: ['地拉那', 'di la na'] } },
  {
    v: 'Europe/Vienna',
    c: { Vienna: ['维也纳', 'wei ye na'], Linz: ['林茨', 'lin ci'] },
  },
  {
    v: 'Europe/Warsaw',
    c: {
      Warsaw: ['华沙', 'hua sha'],
      Katowice: ['卡托维兹', 'ka tuo wei zi'],
      Tshwane: ['茨瓦内', 'ci wa nei'],
    },
  },
  {
    v: 'Europe/Zurich',
    c: {
      Zurich: ['苏黎世', 'su li shi'],
      Geneva: ['日内瓦', 'ri nei wa'],
      Lausanne: ['洛桑', 'luo sang'],
    },
  },
  {
    v: 'Africa/Lagos',
    c: {
      Lagos: ['拉各斯', 'la ge si'],
      Luanda: ['罗安达', 'luo an da'],
      Kinshasa: ['金沙萨', 'jin sha sa'],
    },
  },
  { v: 'Africa/Ndjamena', c: { Ndjamena: ['恩贾梅纳', 'en jia mei na'] } },
  { v: 'Africa/Khartoum', c: { Khartoum: ['喀土穆', 'ka tu mu'] } },
  { v: 'Africa/Maputo', c: { Maputo: ['马普托', 'ma pu tuo'] } },
  { v: 'Africa/Windhoek', c: { Windhoek: ['温得和克', 'wen de he ke'] } },
  { v: 'Africa/Cairo', c: { Cairo: ['开罗', 'kai luo'] } },
  {
    v: 'Europe/Kaliningrad',
    c: { Kaliningrad: ['加里宁格勒', 'jia li ning ge le'] },
  },
  { v: 'Africa/Tripoli', c: { Tripoli: ['的黎波里', 'di li bo li'] } },
  { v: 'Europe/Athens', c: { Athens: ['雅典', 'ya dian'] } },
  { v: 'Asia/Beirut', c: { Beirut: ['贝鲁特', 'bei lu te'] } },
  {
    v: 'Europe/Bucharest',
    c: { Bucharest: ['布加勒斯特', 'bu jia le si te'] },
  },
  { v: 'Europe/Chisinau', c: { Chisinau: ['基希讷乌', 'ji xi ne wu'] } },
  { v: 'Asia/Gaza', c: { Gaza: ['加沙', 'jia sha'] } },
  { v: 'Asia/Hebron', c: { Hebron: ['希伯伦', 'xi bo lun'] } },
  { v: 'Europe/Helsinki', c: { Helsinki: ['赫尔辛基', 'he er xin ji'] } },
  { v: 'Europe/Kiev', c: { Kiev: ['基辅', 'ji fu'] } },
  { v: 'Asia/Nicosia', c: { Nicosia: ['尼科西亚', 'ni ke xi ya'] } },
  { v: 'Europe/Riga', c: { Riga: ['里加', 'li jia'] } },
  { v: 'Europe/Sofia', c: { Sofia: ['索菲亚', 'suo fei ya'] } },
  { v: 'Europe/Tallinn', c: { Tallinn: ['塔林', 'ta lin'] } },
  { v: 'Europe/Uzhgorod', c: { Uzhgorod: ['乌日哥罗德', 'wu ri ge luo de'] } },
  { v: 'Europe/Vilnius', c: { Vilnius: ['维尔纽斯', 'wei er niu si'] } },
  { v: 'Europe/Zaporozhye', c: { Zaporozhye: ['扎波罗热', 'zha bo luo re'] } },
  { v: 'Asia/Famagusta', c: { Famagusta: ['法马古斯塔', 'fa ma gu si ta'] } },
  {
    v: 'Africa/Johannesburg',
    c: {
      CapeTown: ['开普敦', 'kai pu dun'],
      Johannesburg: ['约翰内斯堡', 'yue han nei si bao'],
      Durban: ['德班', 'de ban'],
    },
  },
  { v: 'Africa/Juba', c: { Juba: ['朱巴', 'zhu ba'] } },
  { v: 'Asia/Amman', c: { Amman: ['安曼', 'an man'] } },
  { v: 'Asia/Damascus', c: { Damascus: ['大马士革', 'da ma shi ge'] } },
  { v: 'Asia/Tel_Aviv', c: { TelAviv: ['特拉维夫', 'te la wei fu'] } },
  {
    v: 'Asia/Jerusalem',
    c: { Jerusalem: ['耶路撒冷', 'ye lu sa leng'], Haifa: ['海法', 'hai fa'] },
  },
  { v: 'Asia/Baghdad', c: { Baghdad: ['巴格达', 'ba ge da'] } },
  {
    v: 'Asia/Qatar',
    c: {
      Qatar: ['卡塔尔', 'ka ta er'],
      Doha: ['多哈', 'duo ha'],
      Manama: ['麦纳麦', 'mai na mai'],
    },
  },
  {
    v: 'Asia/Riyadh',
    c: {
      Riyadh: ['利雅得', 'li ya de'],
      Jeddah: ['吉达', 'ji da'],
      KuwaitCity: ['科威特城', 'ke wei te cheng'],
      Dammam: ['达曼', 'da man'],
      Medina: ['麦地那', 'mai de na'],
    },
  },
  {
    v: 'Africa/Nairobi',
    c: { Nairobi: ['内罗毕', 'nei luo bi'], Kampala: ['坎帕拉', 'kan pa la'] },
  },
  { v: 'Europe/Kirov', c: { Kirov: ['基洛夫', 'ji luo fu'] } },
  { v: 'Europe/Minsk', c: { Minsk: ['明斯克', 'ming si ke'] } },
  {
    v: 'Europe/Moscow',
    c: {
      Moscow: ['莫斯科', 'mo si ke'],
      'St.Petersburg': ['圣彼得堡', 'sheng bi de bao'],
      NizhnyNovgorod: ['下诺夫哥罗德', 'xia nuo fu ge luo de'],
      Sochi: ['索契', 'suo qi'],
    },
  },
  {
    v: 'Europe/Simferopol',
    c: { Simferopol: ['辛菲罗波尔', 'xin fei luo bo er'] },
  },
  { v: 'Antarctica/Syowa', c: { SyowaStation: ['昭和站', 'zhao he zhan'] } },
  {
    v: 'Europe/Istanbul',
    c: {
      Istanbul: ['伊斯坦布尔', 'yi si tan bu er'],
      Ankara: ['安卡拉', 'an ka la'],
      Izmir: ['伊兹密尔', 'yi zi mi er'],
    },
  },
  {
    v: 'Europe/Volgograd',
    c: { Volgograd: ['伏尔加格勒', 'fu er jia ge le'] },
  },
  { v: 'Asia/Yerevan', c: { Yerevan: ['埃里温', 'ai li wen'] } },
  { v: 'Europe/Astrakhan', c: { Astrakhan: ['阿斯特拉罕', 'a si te la han'] } },
  { v: 'Asia/Baku', c: { Baku: ['巴库', 'ba ku'] } },
  { v: 'Asia/Tbilisi', c: { Tbilisi: ['第比利斯', 'di bi li si'] } },
  {
    v: 'Asia/Dubai',
    c: {
      Dubai: ['迪拜', 'di bai'],
      AbuDhabi: ['阿布扎比', 'a bu zha bi'],
      Muscat: ['马斯喀特', 'ma si ka te'],
    },
  },
  { v: 'Indian/Mauritius', c: { Mauritius: ['毛里求斯', 'mao li qiu si'] } },
  { v: 'Indian/Reunion', c: { Reunion: ['留尼汪', 'liu ni wang'] } },
  { v: 'Europe/Samara', c: { Samara: ['萨马拉', 'sa ma la'] } },
  { v: 'Europe/Saratov', c: { Saratov: ['萨拉托夫', 'sa la tuo fu'] } },
  { v: 'Indian/Mahe', c: { Seychelles: ['塞舌尔', 'sai she er'] } },
  {
    v: 'Europe/Ulyanovsk',
    c: { Ulyanovsk: ['乌里扬诺夫斯克', 'wu li yang nuo fu si ke'] },
  },
  { v: 'Asia/Tehran', c: { Tehran: ['德黑兰', 'de hei lan'] } },
  { v: 'Asia/Kabul', c: { Kabul: ['喀布尔', 'ka bu er'] } },
  {
    v: 'Indian/Kerguelen',
    c: { KerguelenIslands: ['凯尔盖朗群岛', 'kai er gai lang qun dao'] },
  },
  { v: 'Indian/Maldives', c: { Maldives: ['马尔代夫', 'ma er dai fu'] } },
  { v: 'Antarctica/Mawson', c: { MawsonStation: ['莫森站', 'mo sen zhan'] } },
  {
    v: 'Asia/Karachi',
    c: {
      Islamabad: ['伊斯兰堡', 'yi si lan bao'],
      Karachi: ['卡拉奇', 'ka la qi'],
      Lahore: ['拉合尔', 'la he er'],
    },
  },
  { v: 'Asia/Dushanbe', c: { Dushanbe: ['杜尚别', 'du shang bie'] } },
  { v: 'Asia/Ashgabat', c: { Ashgabat: ['阿什喀巴德', 'a shi ka ba de'] } },
  { v: 'Asia/Samarkand', c: { Samarkand: ['撒马尔罕', 'sa ma er han'] } },
  { v: 'Asia/Tashkent', c: { Tashkent: ['塔什干', 'ta shi gan'] } },
  { v: 'Asia/Aqtau', c: { Aqtau: ['阿克套', 'a ke tao'] } },
  { v: 'Asia/Aqtobe', c: { Aqtobe: ['阿克托别', 'a ke tuo bie'] } },
  { v: 'Asia/Atyrau', c: { Atyrau: ['阿特劳', 'a te lao'] } },
  { v: 'Asia/Oral', c: { Oral: ['乌拉尔', 'wu la er'] } },
  {
    v: 'Asia/Yekaterinburg',
    c: { Yekaterinburg: ['叶卡捷琳堡', 'ye ka jie lin bao'] },
  },
  { v: 'Asia/Qyzylorda', c: { Qyzylorda: ['克孜洛尔达', 'ke zi luo er da'] } },
  { v: 'Asia/Colombo', c: { Colombo: ['科伦坡', 'ke lun po'] } },
  {
    v: 'Asia/Calcutta',
    c: {
      Calcutta: ['加尔各答', 'jia er ge da'],
      Mumbai: ['孟买', 'meng mai'],
      NewDelhi: ['新德里', 'xin de li'],
      Bangalore: ['班加罗尔', 'ban jia luo er'],
      Hyderabad: ['海德拉巴', 'hai de la ba'],
      Chennai: ['金奈', 'jin nai'],
      Ahmedabad: ['艾哈迈达巴德', 'ai ha mai da ba de'],
      Pune: ['浦那', 'pu na'],
      Surat: ['苏拉特', 'su la te'],
      Gurgaon: ['古爾岡', 'gu er gang'],
    },
  },
  { v: 'Asia/Katmandu', c: { Kathmandu: ['加德满都', 'jia de man du'] } },
  { v: 'Asia/Dhaka', c: { Dhaka: ['达卡', 'da ka'] } },
  { v: 'Asia/Thimphu', c: { Thimphu: ['廷布', 'ting bu'] } },
  { v: 'Asia/Almaty', c: { Almaty: ['阿拉木图', 'a la mu tu'] } },
  {
    v: 'Indian/Chagos',
    c: { ChagosArchipelago: ['查戈斯群岛', 'cha ge si qun dao'] },
  },
  { v: 'Asia/Bishkek', c: { Bishkek: ['比什凯克', 'bi shi kai ke'] } },
  { v: 'Asia/Omsk', c: { Omsk: ['鄂木斯克', 'e mu si ke'] } },
  { v: 'Asia/Urumqi', c: { Urumqi: ['乌鲁木齐', 'wu lu mu qi'] } },
  { v: 'Antarctica/Vostok', c: { Vostok: ['沃斯托克', 'wo si tuo ke'] } },
  {
    v: 'Indian/Cocos',
    c: { CocosIslands: ['科科斯群岛', 'ke ke si qun dao'] },
  },
  { v: 'Asia/Yangon', c: { Yangon: ['仰光', 'yang guang'] } },
  { v: 'Asia/Barnaul', c: { Barnaul: ['巴尔瑙尔', 'ba er nao er'] } },
  {
    v: 'Indian/Christmas',
    c: { ChristmasIsland: ['圣诞岛', 'sheng dan dao'] },
  },
  {
    v: 'Antarctica/Davis',
    c: { DavisStation: ['戴维斯站', 'dai wei si zhan'] },
  },
  { v: 'Asia/Hovd', c: { Hovd: ['科布多', 'ke bu duo'] } },
  {
    v: 'Asia/Bangkok',
    c: {
      Bangkok: ['曼谷', 'man gu'],
      ChiangMai: ['清迈', 'qing mai'],
      PhnomPenh: ['金边', 'jin bian'],
    },
  },
  {
    v: 'Asia/Saigon',
    c: {
      HoChiMinhCity: ['胡志明市', 'hu zhi ming shi'],
      Hanoi: ['河内', 'he nei'],
    },
  },
  {
    v: 'Asia/Krasnoyarsk',
    c: { Krasnoyarsk: ['克拉斯诺亚尔斯克', 'ke la si nuo ya er si ke'] },
  },
  {
    v: 'Asia/Novokuznetsk',
    c: { Novokuznetsk: ['新库兹涅茨克', 'xin ku zi nie ci ke'] },
  },
  {
    v: 'Asia/Novosibirsk',
    c: { Novosibirsk: ['新西伯利亚', 'xin xi bo li ya'] },
  },
  { v: 'Asia/Tomsk', c: { Tomsk: ['托木斯克', 'tuo mu si ke'] } },
  {
    v: 'Asia/Jakarta',
    c: {
      Jakarta: ['雅加达', 'ya jia da'],
      Surabaya: ['泗水', 'si shui'],
      Bandung: ['万隆', 'wan long'],
    },
  },
  { v: 'Asia/Pontianak', c: { Pontianak: ['坤甸', 'kun dian'] } },
  { v: 'Australia/Perth', c: { Perth: ['珀斯', 'po si'] } },
  {
    v: 'Asia/Brunei',
    c: { BandarSeriBegawan: ['斯里巴加湾', 'si li ba jia wan'] },
  },
  { v: 'Asia/Makassar', c: { Makassar: ['望加锡', 'wang jia xi'] } },
  { v: 'Asia/Macau', c: { Macao: ['澳门', 'ao men'] } },
  {
    v: 'Asia/Shanghai',
    c: {
      Beijing: ['北京', 'bei jing'],
      Chongqing: ['重庆', 'chong qing'],
      Shanghai: ['上海', 'shang hai'],
      Tianjin: ['天津', 'tian jin'],
      Jinan: ['济南', 'ji nan'],
      Qingdao: ['青岛', 'qing dao'],
      Yantai: ['烟台', 'yan tai'],
      Zhengzhou: ['郑州', 'zheng zhou'],
      Taiyuan: ['太原', 'tai yuan'],
      Linfen: ['临汾', 'lin fen'],
      Shenyang: ['沈阳', 'shen yang'],
      Nanjing: ['南京', 'nan jing'],
      Wuxi: ['无锡', 'wu xi'],
      Suzhou: ['苏州', 'su zhou'],
      Hangzhou: ['杭州', 'hang zhou'],
      Ningbo: ['宁波', 'ning bo'],
      Yiwu: ['义乌', 'yi wu'],
      Wenzhou: ['温州', 'wen zhou'],
      Wuhan: ['武汉', 'wu han'],
      Hefei: ['合肥', 'he fei'],
      Nanchang: ['南昌', 'nan chang'],
      Chengdu: ['成都', 'chéng dū'],
      Neijiang: ['内江', 'nei jiang'],
      Yibin: ['宜宾', 'yi bin'],
      Xian: ['西安', 'xi an'],
      Lanzhou: ['兰州', 'lan zhou'],
      Kunming: ['昆明', 'kun ming'],
      Zhuhai: ['珠海', 'zhu hai'],
      Shenzhen: ['深圳', 'shen zhen'],
      Dongguan: ['东莞', 'dong guan'],
      Foshan: ['佛山', 'fo shan'],
      Guangzhou: ['广州', 'guang zhou'],
      Qingyuan: ['清远', 'qing yuan'],
      Fuzhou: ['福州', 'fu zhou'],
      Xiamen: ['厦门', 'xia men'],
      Longyan: ['龙岩', 'long yan'],
      Changsha: ['长沙', 'chang sha'],
      Haikou: ['海口', 'hai kou'],
      Sanya: ['三亚', 'san ya'],
      Chengmai: ['澄迈', 'cheng mai'],
    },
  },
  { v: 'Asia/Choibalsan', c: { Choibalsan: ['乔巴山', 'qiao ba shan'] } },
  { v: 'Asia/Hong_Kong', c: { Hongkong: ['香港', 'xiang gang'] } },
  { v: 'Asia/Irkutsk', c: { Irkutsk: ['伊尔库茨克', 'yi er ku ci ke'] } },
  {
    v: 'Asia/Kuala_Lumpur',
    c: {
      KualaLumpur: ['吉隆坡', 'ji long po'],
      Chittagong: ['吉大港', 'ji da gang'],
    },
  },
  { v: 'Asia/Kuching', c: { Kuching: ['古晋', 'gu jin'] } },
  { v: 'Asia/Manila', c: { Manila: ['马尼拉', 'ma ni la'] } },
  { v: 'Asia/Singapore', c: { Singapore: ['新加坡', 'xin jia po'] } },
  { v: 'Asia/Taipei', c: { Taipei: ['台北', 'tai bei'] } },
  { v: 'Asia/Ulaanbaatar', c: { Ulaanbaatar: ['乌兰巴托', 'wu lan ba tuo'] } },
  { v: 'Australia/Eucla', c: { Eucla: ['尤克拉', 'you ke la'] } },
  { v: 'Asia/Dili', c: { Dili: ['帝力', 'di li'] } },
  { v: 'Asia/Jayapura', c: { Jayapura: ['查亚普拉', 'cha ya pu la'] } },
  {
    v: 'Asia/Tokyo',
    c: {
      Tokyo: ['东京', 'dong jing'],
      Osaka: ['大阪', 'da ban'],
      Nagoya: ['名古屋', 'ming gu wu'],
      Yokohama: ['横滨', 'heng bin'],
      Kyoto: ['京都', 'jing du'],
      Fukuoka: ['福冈', 'fu gang'],
      Sapporo: ['札幌', 'zha huang'],
      Kobe: ['神户', 'shen hu'],
      Saitama: ['埼玉', 'qi yu'],
      Okinawa: ['冲绳', 'chong sheng'],
      Fuji: ['富士', 'fu shi'],
      Hiroshima: ['广岛', 'guang dao'],
      Nagasaki: ['长崎', 'chang qi'],
      Chiba: ['千叶', 'qian ye'],
      Sendai: ['仙台', 'xian tai'],
      Kawasaki: ['川崎', 'chuan qi'],
      Okayama: ['冈山', 'gang shan'],
      Nagano: ['长野', 'chang ye'],
    },
  },
  { v: 'Asia/Pyongyang', c: { Pyongyang: ['平壤', 'ping rang'] } },
  {
    v: 'Asia/Seoul',
    c: {
      Seoul: ['首尔', 'shou er'],
      Busan: ['釜山', 'fu shan'],
      Incheon: ['仁川', 'ren chuan'],
      Ulsan: ['蔚山', 'wei shan'],
      Changwon: ['昌原', 'chang yuan'],
      Daegu: ['大邱', 'da qiu'],
    },
  },
  { v: 'Pacific/Palau', c: { Palau: ['帕劳', 'pa lao'] } },
  { v: 'Asia/Chita', c: { Chita: ['赤塔', 'chi ta'] } },
  { v: 'Asia/Khandyga', c: { Khandyga: ['汉德加', 'han de jia'] } },
  { v: 'Asia/Yakutsk', c: { Yakutsk: ['雅库茨克', 'ya ku ci ke'] } },
  { v: 'Australia/Darwin', c: { Darwin: ['达尔文', 'da er wen'] } },
  {
    v: 'Australia/Brisbane',
    c: {
      Brisbane: ['布里斯班', 'bu li si ban'],
      GoldCoast: ['黄金海岸', 'huang jin hai an'],
    },
  },
  { v: 'Australia/Lindeman', c: { LindemanIsland: ['林德曼', 'lin de man'] } },
  { v: 'Pacific/Guam', c: { Chamorro: ['查莫罗', 'cha mo luo'] } },
  { v: 'Pacific/Chuuk', c: { Chuuk: ['楚克', 'chu ke'] } },
  {
    v: 'Antarctica/DumontDUrville',
    c: {
      DumontDUrvilleStation: ['迪蒙迪尔维尔站', 'di meng di er wei er zhan'],
    },
  },
  {
    v: 'Pacific/Port_Moresby',
    c: { PortMoresby: ['莫尔斯比港', 'mo er si bi gang'] },
  },
  { v: 'Asia/Ust-Nera', c: { 'Ust-Nera': ['乌斯内拉', 'wu si nei la'] } },
  {
    v: 'Asia/Vladivostok',
    c: { Vladivostok: ['符拉迪沃斯托克', 'fu la di wo si tuo ke'] },
  },
  { v: 'Australia/Adelaide', c: { Adelaide: ['阿德莱德', 'a de lai de'] } },
  {
    v: 'Australia/Broken_Hill',
    c: { BrokenHill: ['布罗肯希尔', 'bu luo ken xi er'] },
  },
  { v: 'Antarctica/Casey', c: { CaseyStation: ['凯西站', 'kai xi zhan'] } },
  {
    v: 'Pacific/Bougainville',
    c: { Bougainville: ['布干维尔', 'bu gan wei er'] },
  },
  { v: 'Australia/Currie', c: { Currie: ['库利', 'ku li'] } },
  { v: 'Australia/Hobart', c: { Hobart: ['霍巴特', 'huo ba te'] } },
  { v: 'Australia/Melbourne', c: { Melbourne: ['墨尔本', 'mo er ben'] } },
  {
    v: 'Australia/Sydney',
    c: { Sydney: ['悉尼', 'xi ni'], Canberra: ['堪培拉', 'kan pei la'] },
  },
  { v: 'Pacific/Kosrae', c: { Kosrae: ['科斯雷', 'ke si lei'] } },
  {
    v: 'Australia/Lord_Howe',
    c: { LordHoweIsland: ['豪勋爵岛', 'hao xun jue dao'] },
  },
  {
    v: 'Antarctica/Macquarie',
    c: { MacquarieIsland: ['麦夸里岛', 'mai kua li dao'] },
  },
  { v: 'Asia/Magadan', c: { Magadan: ['马加丹', 'ma jia dan'] } },
  {
    v: 'Pacific/Noumea',
    c: { NewCaledonia: ['新喀里多尼亚', 'xin ka li duo ni ya'] },
  },
  { v: 'Pacific/Pohnpei', c: { Pohnpei: ['波纳佩', 'bo na pei'] } },
  { v: 'Asia/Sakhalin', c: { Sakhalin: ['库页岛', 'ku ye dao'] } },
  {
    v: 'Pacific/Guadalcanal',
    c: { SolomonIslands: ['所罗门群岛', 'suo luo men qun dao'] },
  },
  {
    v: 'Asia/Srednekolymsk',
    c: { Srednekolymsk: ['中科雷姆斯克', 'zhong ke lei mu si ke'] },
  },
  { v: 'Pacific/Efate', c: { Efate: ['埃法特岛', 'ai fa te dao'] } },
  { v: 'Pacific/Norfolk', c: { NorfolkIsland: ['诺福克岛', 'nuo fu ke dao'] } },
  { v: 'Asia/Anadyr', c: { Anadyr: ['阿纳德尔', 'a na de er'] } },
  {
    v: 'Pacific/Tarawa',
    c: { GilbertIslands: ['吉尔伯特群岛', 'ji er bo te qun dao'] },
  },
  { v: 'Pacific/Kwajalein', c: { Kwajalein: ['夸贾林', 'kua jia lin'] } },
  { v: 'Pacific/Majuro', c: { Majuro: ['马朱罗', 'ma zhu luo'] } },
  { v: 'Pacific/Nauru', c: { Nauru: ['瑙鲁', 'nao lu'] } },
  {
    v: 'Asia/Kamchatka',
    c: {
      PetropavlovskKamchatski: ['堪察加彼得巴甫洛夫斯克', 'kan cha jia bi de ba fu luo fu si ke'],
    },
  },
  { v: 'Pacific/Funafuti', c: { Tuvalu: ['图瓦卢', 'tu wa lu'] } },
  { v: 'Pacific/Wake', c: { WakeIsland: ['威克岛', 'wei ke dao'] } },
  { v: 'Pacific/Wallis', c: { MataUtu: ['马塔乌图', 'ma ta wu tu'] } },
  { v: 'Pacific/Fiji', c: { Fiji: ['斐济', 'fei ji'] } },
  {
    v: 'Pacific/Auckland',
    c: {
      Wellington: ['惠灵顿', 'hui ling dun'],
      Auckland: ['奥克兰', 'ao ke lan'],
      Christchurch: ['基督城', 'ji du cheng'],
    },
  },
  {
    v: 'Pacific/Enderbury',
    c: { PhoenixIslands: ['菲尼克斯群岛', 'fei ni ke si qun dao'] },
  },
  { v: 'Pacific/Fakaofo', c: { Tokelau: ['托克劳', 'tuo ke lao'] } },
  {
    v: 'Pacific/Tongatapu',
    c: { NukuAlofa: ['努库阿洛法', 'nu ku a luo fa'] },
  },
  { v: 'Pacific/Apia', c: { Apia: ['阿皮亚', 'a pi ya'] } },
  {
    v: 'Pacific/Chatham',
    c: { ChathamIslands: ['查塔姆群岛', 'cha ta mu qun dao'] },
  },
  {
    v: 'Pacific/Kiritimati',
    c: { LineIslands: ['莱恩群岛', 'lai en qun dao'] },
  },
];
