// @universe-design/react/src/pro/timezone-picker/types.ts
import type React from 'react';
import type moment from '@lark/moment-timezone';
import type { SingleSelectProps } from '../../select/interface';
import type locales from './locales/en-US';
import type { IBuiltinEntries, ISearchEntries } from './raw-data';

export type { ISearchEntries } from './raw-data';

type Locales = typeof locales;
type LocaleKey = keyof Locales;

export interface TimezonePickerProps
  extends Omit<SingleSelectProps<string>, 'filterOption' | 'searchValue' | 'onSearch' | 'options'> {
  date?: moment.Moment;
  /**
   * @experimental
   */
  metaTransformer?: TimezoneMetaTransformer;
  /**
   * @deprecated locale 现在不会报错，默认使用 enUS 兜底
   */
  onError?: (err: Error) => void;
}

/**
 * 处于压缩目的，尽量减少数据内容
 * @internal
 */
export interface TimeZoneRawMeta {
  /** 时区 ID */
  v: TimezoneValue;
  /** 城市信息 */
  c: Partial<Record<CityLocaleKey, IBuiltinEntries>>;
}

export interface TimeZoneMeta extends TimeZoneRawMeta {
  /** 时区名称 */
  l: string;
  s: ISearchEntries[];
}

/** `Timezone` 值 */
export type TimezoneValue = string;
/** `Timezone` 国际化文案类型 */
export type TimezoneLocale = Record<TimezoneValue, string>;
/** `Timezone` 旧数据值 */
export type LegacyTimezoneValue = string;
/** `Timezone` 国际化文案键 */
export type TimezoneLocaleKey = Extract<LocaleKey, `Tz_${string}`>;
type ExtractCityLocaleKey<T> = T extends `Ct_${infer U}` ? U : never;
/** `City` 国际化文案键 */
export type CityLocaleKey = ExtractCityLocaleKey<LocaleKey>;

/**
 * `Timezone` 格式化数据
 */
export type TimezoneFormatData = {
  /**
   * GMT 偏移量
   * @example 'GMT+08:00'
   */
  GMTOffset: string;
  /**
   * 经过国际化转换的区名称
   * @example '中国标准时间'
   */
  timeName: string;
  /**
   * 经过国际化转换的区名称
   * @example '中国标准时间 - 北京'
   */
  zoneName: string;
  /**
   * 时区名称是否唯一
   */
  isUnique: boolean;
  /** 城市名称 */
  cities: string[];
};

/**
 * `TimezoneLabel` 格式化函数
 */
export type TimezoneLabelFormatter = (data: TimezoneFormatData | undefined) => React.ReactNode;

export type TimezoneMetaTransformer = (
  meta: TimeZoneMeta,
  metaMap: Record<TimezoneValue, TimeZoneMeta>,
) => TimeZoneMeta;
