// @universe-design/react/src/card/stories/homeless.stories.tsx
import React, { useState } from 'react';
import type { RadioValue, DefaultSizeType } from '../..';
import { Card, Switch, Divider, Row, Col, Radio, DefaultSizeEnum, Text } from '../..';
import stylex from '@universe-design/stylex';

import './index.less';

export default {
  title: 'Components/Card/临时示例', // 注意：这里需要修改成自己对应 Component name。
  component: Card,
};

const body = () => (
  <div>
    <div>This is Card body</div>
    <div>This is Card body</div>
    <div>This is Card body</div>
    <div>This is Card body</div>
    <div>This is Card body</div>
  </div>
);

export { default as Basic } from '../demo/basic';
export { default as Border } from '../demo/border';
export { default as BorderRadius } from '../demo/border-radius';
export { default as NoHoverState } from '../demo/no-hover-state';
export { default as Shadow } from '../demo/shadow';
export { default as ShadowSize } from '../demo/shadow-size';

export const BaseCase = () => {
  const [bordered, setBordered] = useState(true);
  const [shadowed, setShadowed] = useState(true);
  const [hoverState, setHoverState] = useState(true);

  const [shadowType, setShadowType] = useState<DefaultSizeType>('medium');
  const [brType, setBrType] = useState<DefaultSizeType>('small');

  const handleShadowTypeChange = (value?: RadioValue) => setShadowType(value as any);

  const handleBrTypeChange = (value?: RadioValue) => setBrType(value as any);

  return (
    <div className="ud-card-demo">
      <div className="sw">
        <span className="l">边框: </span>
        <Switch checked={bordered} onChange={setBordered} />
      </div>
      <div className="sw">
        <span className="l">阴影: </span>
        <Switch checked={shadowed} onChange={setShadowed} />
      </div>
      <div className="sw">
        <span className="l">hover 状态: </span>
        <Switch checked={hoverState} onChange={setHoverState} />
      </div>
      <div className="sw">
        <span className="l">阴影大小: </span>
        <Radio.Group onChange={handleShadowTypeChange} value={shadowType}>
          <Radio.Button value="large">大</Radio.Button>
          <Radio.Button value="medium">中</Radio.Button>
          <Radio.Button value="small">小</Radio.Button>
        </Radio.Group>
      </div>
      <div className="sw">
        <span className="l">圆角大小: </span>
        <Radio.Group onChange={handleBrTypeChange} value={brType}>
          <Radio.Button value={DefaultSizeEnum[0]}>大</Radio.Button>
          <Radio.Button value={DefaultSizeEnum[1]}>中</Radio.Button>
          <Radio.Button value={DefaultSizeEnum[2]}>小</Radio.Button>
        </Radio.Group>
      </div>
      <Divider text="基础" />
      <Card
        shadowed={shadowed}
        bordered={bordered}
        shadow={shadowType}
        borderRadius={brType}
        noHoverState={!hoverState}
      >
        {body()}
      </Card>
      <Divider text="自定义宽度" />
      <Card
        style={{ width: 400 }}
        shadowed={shadowed}
        bordered={bordered}
        shadow={shadowType}
        borderRadius={brType}
        noHoverState={!hoverState}
      >
        {body()}
      </Card>
      <Divider text="与栅格结合" />
      <Row gutter={[16, 16]}>
        <Col span={6}>
          <Card
            shadowed={shadowed}
            bordered={bordered}
            shadow={shadowType}
            borderRadius={brType}
            noHoverState={!hoverState}
          >
            {body()}
          </Card>
        </Col>
        <Col span={6}>
          <Card
            shadowed={shadowed}
            bordered={bordered}
            shadow={shadowType}
            borderRadius={brType}
            noHoverState={!hoverState}
          >
            {body()}
          </Card>
        </Col>
        <Col span={6}>
          <Card
            shadowed={shadowed}
            bordered={bordered}
            shadow={shadowType}
            borderRadius={brType}
            noHoverState={!hoverState}
          >
            {body()}
          </Card>
        </Col>
        <Col span={6}>
          <Card
            shadowed={shadowed}
            bordered={bordered}
            shadow={shadowType}
            borderRadius={brType}
            noHoverState={!hoverState}
          >
            {body()}
          </Card>
        </Col>
        <Divider text="点击事件" />
        <Col span={6}>
          <Card onClick={() => alert('clicked')}>{body()}</Card>
        </Col>
      </Row>
    </div>
  );
};

export const HalfPixel = () => {
  const DivDivider = React.useCallback(
    ({ isHalf }: { isHalf: boolean }) => (
      <div
        className={stylex(
          stylex.create({
            backgroundColor: $token['line-divider-default'],
          }),
        )}
        style={{
          width: isHalf ? '200%' : '100%',
          height: 1,
          transform: isHalf ? 'scale(0.5)' : 'none',
        }}
      />
    ),
    [],
  );

  return (
    <>
      <Text variant="title3">0.5px 实现采用 css 缩放属性 transform - scale 实现</Text>
      <br />
      <Text>
        水平方向和垂直方向，通过 -50%，同时扩大一倍，width 和 height 变成原来两倍，然后 scale(0.5) 缩小为原始尺寸；
      </Text>
      <br />
      <br />
      <div
        className={stylex(
          stylex.create({
            boxSizing: 'border-box',
            padding: 12,
            backgroundColor: $token['bg-body'],
            maxWidth: 400,
            position: 'relative',

            '::before': {
              content: '""',
              position: 'absolute',
              inset: '-50%',
              borderRadius: 16,
              borderWidth: 1,
              borderStyle: 'solid',
              borderColor: $token['line-border-card'],
              transform: 'scale(0.5)',
            },
          }),
        )}
      >
        <Text variant="headline">0.5px border & divider</Text>
        <div
          className={stylex(
            stylex.create({
              display: 'grid',
              gap: 8,
              gridTemplateColumns: '1fr 1fr 1fr',
              alignItems: 'center',
              justifyItems: 'center',
            }),
          )}
        >
          <DivDivider isHalf />
          <Text>0.5px div</Text>
          <DivDivider isHalf />
        </div>
      </div>
      <br />
      <div
        className={stylex(
          stylex.create({
            boxSizing: 'border-box',
            padding: 12,
            borderRadius: 8,
            backgroundColor: $token['bg-body'],
            borderWidth: 1,
            borderStyle: 'solid',
            borderColor: $token['line-border-card'],
            maxWidth: 400,
          }),
        )}
      >
        <Text variant="headline">1px border & divider</Text>
        <div
          className={stylex(
            stylex.create({
              display: 'grid',
              gap: 8,
              gridTemplateColumns: '1fr 1fr 1fr',
              alignItems: 'center',
              justifyItems: 'center',
            }),
          )}
        >
          <DivDivider isHalf={false} />
          <Text>1px div</Text>
          <DivDivider isHalf={false} />
        </div>
      </div>
    </>
  );
};
