// @universe-design/react/src/card/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import CBasic from '../demo/basic';
import CBorderRadius from '../demo/border-radius';
import CBorder from '../demo/border';
import CNoHoverState from '../demo/no-hover-state';
import CShadowSize from '../demo/shadow-size';
import CShadow from '../demo/shadow';
if (!(CBasic as Story).storyName) {
  (CBasic as Story).storyName = '0. 基础用法';
  (CBasic as Story).parameters = {
    docs: {
      description: {
        story: '简单的用法。\n\n',
      },
    },
  };
}
if (!(CBorderRadius as Story).storyName) {
  (CBorderRadius as Story).storyName = '4. 卡片圆角';
  (CBorderRadius as Story).parameters = {
    docs: {
      description: {
        story: '通过`borderRadius`设置卡片圆角的大小。共有 2 种尺寸，分别为`large`、`small`，默认尺寸为`small`。\n\n',
      },
    },
  };
}
if (!(CBorder as Story).storyName) {
  (CBorder as Story).storyName = '1. 卡片边框';
  (CBorder as Story).parameters = {
    docs: {
      description: {
        story: '通过`bordered`属性，可以控制是否渲染边框。\n\n',
      },
    },
  };
}
if (!(CNoHoverState as Story).storyName) {
  (CNoHoverState as Story).storyName = '5. 无鼠标交互';
  (CNoHoverState as Story).parameters = {
    docs: {
      description: {
        story: '通过`noHoverState`属性，可以关闭卡片的鼠标`hover`交互效果。\n\n',
      },
    },
  };
}
if (!(CShadowSize as Story).storyName) {
  (CShadowSize as Story).storyName = '3. 卡片阴影大小';
  (CShadowSize as Story).parameters = {
    docs: {
      description: {
        story:
          '通过`shadow`属性，可以控制卡片阴影的大小。共有 3 种尺寸，分别为`large`、`medium`、`small`，默认尺寸为`medium`。\n\n',
      },
    },
  };
}
if (!(CShadow as Story).storyName) {
  (CShadow as Story).storyName = '2. 卡片阴影';
  (CShadow as Story).parameters = {
    docs: {
      description: {
        story: '通过`shadowed`属性，可以控制是否渲染阴影。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Card/官网示例',
  tags: ['autodocs'],
};

export { CBasic, CBorderRadius, CBorder, CNoHoverState, CShadowSize, CShadow };
