// @universe-design/react/es/metrics/MetricsSecondaryContent.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.u112evzf{color:@datav-function-element-neutral;}
import u1xg5upn from "./../.stylex/u1xg5upn";
//.u19ci1c7{color:@datav-function-element-warning;}
import u1amwvao from "./../.stylex/u1amwvao";
//.uy5pp46{color:@datav-function-element-info;}
import ucy8jds from "./../.stylex/ucy8jds";
//.u1bxglv9{color:@datav-function-element-danger;}
import uv1n46n from "./../.stylex/uv1n46n";
//.u1o67kls{color:@datav-function-element-success;}
import u894z0b from "./../.stylex/u894z0b";
//.u1hzsgrp{margin-left:2px;}
import u1agd5ph from "./../.stylex/u1agd5ph";
//.ulkuc6h{line-height:100%;}
import u1e9he2z from "./../.stylex/u1e9he2z";
//.ueyvcx7{font-family:Roboto;}
import umbknfa from "./../.stylex/umbknfa";
//.u1rvp7gu{color:@text-caption;}
import ueh72r3 from "./../.stylex/ueh72r3";
//.u8zmjen{margin-right:4px;}
import u130iqk6 from "./../.stylex/u130iqk6";
//.u14t3ns0{display:inline-block;}
import u8wgnjj from "./../.stylex/u8wgnjj";
//.uz5stix{white-space:nowrap;}
import uugkpzq from "./../.stylex/uugkpzq";
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx, makeStylexStyles } from '@universe-design/stylex/react';
import ExpandDownFilled from '@universe-design/icons-react/ExpandDownFilled';
import ExpandUpFilled from '@universe-design/icons-react/ExpandUpFilled';
import stylex from '@universe-design/stylex';
import React, { useCallback, useState } from 'react';
import { Divider } from '../divider';
import { Text } from '../text';
import { ResizeObserver, useMergedRefs } from '@universe-design/react-utilities';
import { forwardHTMLRef } from '../util/forwardRef';
import { useMetricsContext } from './MetricsContext';
import { presetColor } from './constant';
import { useMetricsSecondaryDescendant } from './useMetrics';
import { useMetricsGroupContextSelector } from './MetricsGroup';
import clsx from 'clsx';
function isPresetColor(color) {
  return presetColor.includes(color);
}
const SecondaryItem = _ref => {
  let {
    index,
    setItemWidths,
    item
  } = _ref;
  return jsx(ResizeObserver, {
    onResize: _ref2 => {
      let {
        width
      } = _ref2;
      setItemWidths(prev => {
        const newWidths = [...prev];
        newWidths[index] = width;
        return newWidths;
      });
    }
  }, jsx(MetricsSecondaryContentItem, _extends({
    ref: useCallback(node => {
      if (!node) {
        setItemWidths(prev => {
          const newWidths = [...prev];
          newWidths.splice(index, 1);
          return newWidths;
        });
      }
    }, [index, setItemWidths])
  }, item)));
};
export const MetricsSecondaryContent = forwardHTMLRef((props, ref) => {
  const {
    secondary,
    className,
    ...restProps
  } = props;
  const secondaryInfo = secondary === null || secondary === void 0 ? void 0 : secondary.data;
  const [itemWidths, setItemWidths] = useState([]);
  const [containerWidth, setContainerWidth] = useState(0);
  const [dividerWidth, setDividerWidth] = useState(0);
  const {
    index,
    register
  } = useMetricsSecondaryDescendant();
  const recalculateSecondaryHeights = useMetricsGroupContextSelector(context => context === null || context === void 0 ? void 0 : context.recalculateSecondaryHeights);
  const secondaryRowMaxHeight = useMetricsGroupContextSelector(context => {
    var _context$secondaryRow, _context$secondaryRow2;
    return context === null || context === void 0 ? void 0 : (_context$secondaryRow = context.secondaryRowInfo) === null || _context$secondaryRow === void 0 ? void 0 : (_context$secondaryRow2 = _context$secondaryRow[index]) === null || _context$secondaryRow2 === void 0 ? void 0 : _context$secondaryRow2.maxHeight;
  });
  const {
    rootCssScope
  } = useMetricsContext();
  const styles = useMetricsSecondaryItemStylexStyles({
    prefix: rootCssScope()
  });
  const layoutSingleRow = React.useCallback((items, dividerWidth, startIndex) => {
    const row = [];
    let totalWidth = itemWidths[startIndex];

    // 添加第一个元素
    row.push({
      item: items[startIndex],
      isLast: false,
      marginRight: 0,
      width: itemWidths[startIndex]
    });
    let currentIndex = startIndex + 1;

    // 尝试添加更多元素
    while (currentIndex < items.length) {
      const nextItemWidth = dividerWidth + itemWidths[currentIndex];
      const newTotalWidth = totalWidth + nextItemWidth;
      if (newTotalWidth > containerWidth) {
        // 当前行放不下更多元素
        const lastItem = row[row.length - 1];
        lastItem.isLast = true;
        lastItem.marginRight = Math.floor(containerWidth - totalWidth) || 0;
        break;
      }

      // 可以继续添加元素
      row.push({
        item: items[currentIndex],
        isLast: false,
        marginRight: 0,
        width: itemWidths[currentIndex]
      });
      totalWidth = newTotalWidth;
      currentIndex++;
    }

    // 处理最后一行
    if (currentIndex === items.length) {
      const lastItem = row[row.length - 1];
      lastItem.isLast = true;
      lastItem.marginRight = Math.floor(containerWidth - totalWidth) || 0;
    }
    return row;
  }, [containerWidth, itemWidths]);
  const secondaryList = React.useMemo(() => {
    if (!secondaryInfo) {
      return [];
    }
    const result = [];
    let currentIndex = 0;
    while (currentIndex < secondaryInfo.length) {
      const row = layoutSingleRow(secondaryInfo, dividerWidth, currentIndex);
      result.push(...row);
      currentIndex += row.length;
    }
    return result;
  }, [secondaryInfo, layoutSingleRow, dividerWidth]);
  if (!secondaryInfo) {
    return null;
  }
  return jsx(React.Fragment, null, jsx("div", _extends({
    className: clsx(`${rootCssScope()}__secondary__wrapper`, className),
    style: {
      height: secondaryRowMaxHeight
    }
  }, restProps), jsx(ResizeObserver, {
    onResize: _ref3 => {
      let {
        width
      } = _ref3;
      setContainerWidth(width);
      recalculateSecondaryHeights === null || recalculateSecondaryHeights === void 0 ? void 0 : recalculateSecondaryHeights();
    }
  }, jsx("div", {
    ref: useMergedRefs(ref, register)
  }, secondaryList.map((_ref4, i) => {
    let {
      item,
      isLast,
      marginRight
    } = _ref4;
    return jsx(React.Fragment, {
      key: i
    }, jsx(SecondaryItem, {
      index: i,
      item: item,
      setItemWidths: setItemWidths
    }), isLast && marginRight > 0 && jsx("span", {
      style: {
        width: marginRight
      },
      sx: [styles.spacer]
    }), !isLast && jsx(ResizeObserver, {
      onResize: _ref5 => {
        let {
          width
        } = _ref5;
        setDividerWidth(width);
      }
    }, jsx("span", {
      style: {
        display: 'inline-block'
      }
    }, jsx(Divider, {
      type: "vertical"
    }))));
  })))));
});
const useMetricsSecondaryItemStylexStyles = makeStylexStyles(_ref6 => {
  let {
    prefix
  } = _ref6;
  return {
    root: {
      "props": {
        "mc9l5x": u8wgnjj,
        "Huce71": uugkpzq
      },
      "alias": `${prefix}__root`
    },
    spacer: {
      "props": {
        "mc9l5x": u8wgnjj
      },
      "alias": `${prefix}__spacer`
    },
    label: {
      "props": {
        "sj55zd": ueh72r3,
        "t21cq0": u130iqk6
      },
      "alias": `${prefix}__label`
    },
    icon: {
      "props": {
        "Bahqtrf": umbknfa
      },
      "alias": `${prefix}__icon`
    },
    value: {
      "props": {
        "Frg6f3": u1agd5ph,
        "Bahqtrf": umbknfa,
        "Bg96gwp": u1e9he2z
      },
      "alias": `${prefix}__value`
    },
    success: {
      "props": {
        "sj55zd": u894z0b
      },
      "alias": `${prefix}__success`
    },
    danger: {
      "props": {
        "sj55zd": uv1n46n
      },
      "alias": `${prefix}__danger`
    },
    info: {
      "props": {
        "sj55zd": ucy8jds
      },
      "alias": `${prefix}__info`
    },
    warning: {
      "props": {
        "sj55zd": u1amwvao
      },
      "alias": `${prefix}__warning`
    },
    neutral: {
      "props": {
        "sj55zd": u1xg5upn
      },
      "alias": `${prefix}__neutral`
    }
  };
});
const MetricsSecondaryContentItem = forwardHTMLRef((props, ref) => {
  const {
    label,
    value: valueNode,
    type = 'none',
    valueStyle,
    color = 'neutral',
    ...restProps
  } = props;
  const {
    size,
    rootCssScope
  } = useMetricsContext();
  const iconMap = {
    increase: jsx(ExpandUpFilled, null),
    decrease: jsx(ExpandDownFilled, null),
    none: null
  };
  const iconNode = iconMap[type];
  const isPreset = isPresetColor(color);
  const styles = useMetricsSecondaryItemStylexStyles({
    prefix: rootCssScope()
  });
  return jsx("span", _extends({
    ref: ref,
    sx: [styles.root]
  }, restProps), label !== null && jsx(Text, {
    variant: "body0",
    sx: [styles.label]
  }, label), iconNode && jsx(Text, {
    sx: [styles.icon, isPreset ? styles[color] : null],
    style: {
      color: isPreset ? undefined : color,
      ...valueStyle
    },
    variant: "headline"
  }, iconNode), valueNode !== null && jsx(Text, {
    sx: [styles.value, isPreset ? styles[color] : null],
    style: {
      color: isPreset ? undefined : color,
      ...valueStyle
    },
    variant: size === 'medium' ? 'headline' : 'title4'
  }, valueNode));
});
