// @universe-design/react/es/metrics/MetricsRoot.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.usbu5mz{gap:12px;}
import u40u48c from "./../.stylex/u40u48c";
//.u4xv25i{gap:4px;}
import umue7z from "./../.stylex/umue7z";
//.u1vx9l62{flex-direction:column;}
import uohwdt from "./../.stylex/uohwdt";
//.u5otk8j{border-top-left-radius:@border-radius-lg;}
import u1d70x8k from "./../.stylex/u1d70x8k";
//.u9kun4h{border-top-right-radius:@border-radius-lg;}
import u1cpp6s6 from "./../.stylex/u1cpp6s6";
//.u1a4u8w2{border-bottom-right-radius:@border-radius-lg;}
import uyretdy from "./../.stylex/uyretdy";
//.uiy6t15{border-bottom-left-radius:@border-radius-lg;}
import unqnxb from "./../.stylex/unqnxb";
//.u22iagw{display:flex;}
import ukchsq7 from "./../.stylex/ukchsq7";
//.u122n59{align-items:center;}
import ubyhozq from "./../.stylex/ubyhozq";
//.u19gb1f4{gap:8px;}
import uelbtaz from "./../.stylex/uelbtaz";
//.uqerorx{flex-grow:1;}
import u99gbyh from "./../.stylex/u99gbyh";
//.uly5x3f{width:100%;}
import u13d3rh8 from "./../.stylex/u13d3rh8";
/* eslint-disable react-hooks/rules-of-hooks */
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx, makeStylexStyles } from '@universe-design/stylex/react';
import { MetricsHeader } from './MetricsHeader';
import { MetricsMainContent } from './MetricsMainContent';
import { MetricsChart } from './MetricsChart';
import { MetricsSecondaryContent } from './MetricsSecondaryContent';
import stylex from '@universe-design/stylex';
import { useMetricsContext } from './MetricsContext';
import { forwardHTMLRef } from '../internal';
import * as React from 'react';
import { useMergedRefs, ResizeObserver } from '@universe-design/react-utilities';
import { useMetricsDescendant } from './useMetrics';
import { useMetricsGroupContextSelector } from './MetricsGroup';
import { MetricsFooter } from './MetricsFooter';
const sxMetricsValue = {
  root: {
    "props": {
      "Bh6795r": u99gbyh,
      "a9b677": u13d3rh8
    }
  }
};
const MetricsValue = forwardHTMLRef((props, ref) => jsx("div", {
  ref: ref,
  className: stylex(sxMetricsValue.root)
}, props.children));
const useMetricsStylexStyles = makeStylexStyles(_ref => {
  let {
    prefix
  } = _ref;
  return {
    root: {
      "props": {
        "Btl43ni": u1d70x8k,
        "B7oj6ja": u1cpp6s6,
        "Bbmb7ep": uyretdy,
        "Beyfa6y": unqnxb,
        "mc9l5x": ukchsq7,
        "Bt984gj": ubyhozq,
        "rmohyg": uelbtaz
      },
      "alias": `${prefix}__root`
    },
    main: {
      "props": {
        "rmohyg": umue7z,
        "mc9l5x": ukchsq7,
        "Beiy3e4": uohwdt
      },
      "alias": `${prefix}__main`
    },
    'secondary-root': {
      "props": {
        "mc9l5x": ukchsq7,
        "Beiy3e4": uohwdt
      },
      "alias": `${prefix}__secondary-root`
    },
    'secondary-medium': {
      "props": {
        "rmohyg": uelbtaz
      },
      "alias": `${prefix}__secondary-medium`
    },
    'secondary-large': {
      "props": {
        "rmohyg": u40u48c
      },
      "alias": `${prefix}__secondary-large`
    }
  };
});
export const MetricsRoot = /*#__PURE__*/React.memo(forwardHTMLRef((props, ref) => {
  const {
    mainTitle,
    mainInfoIcon,
    mainHeaderContentAfter,
    mainValue,
    secondary,
    footer,
    chart,
    ...restProps
  } = props;
  const {
    register
  } = useMetricsDescendant();
  const {
    rootCssScope,
    size,
    position
  } = useMetricsContext();
  const recalculateLocation = useMetricsGroupContextSelector(context => context === null || context === void 0 ? void 0 : context.recalculateLocation);
  const styles = useMetricsStylexStyles({
    prefix: rootCssScope()
  });
  return jsx(ResizeObserver, {
    onResize: recalculateLocation
  }, jsx("div", _extends({
    ref: useMergedRefs(ref, register),
    sx: [styles.root]
  }, restProps), jsx(MetricsValue, null, jsx("div", {
    sx: [styles[`secondary-${size}`], styles['secondary-root']]
  }, jsx("div", {
    sx: [styles.main]
  }, jsx(MetricsHeader, {
    mainTitle: mainTitle,
    mainInfoIcon: mainInfoIcon,
    mainHeaderContentAfter: mainHeaderContentAfter
  }), jsx(MetricsMainContent, {
    mainValue: mainValue,
    chart: chart
  })), jsx(MetricsSecondaryContent, {
    secondary: secondary
  })), jsx(MetricsFooter, {
    footer: footer
  })), position === 'afterAll' && jsx(MetricsChart, {
    chart: chart
  })));
}));
