// @universe-design/react/es/metrics/MetricsGroup.js
import _extends from "@babel/runtime/helpers/esm/extends";
import { createContextHook } from '@universe-design/react-provider';
import { createContext, useContext, useContextSelector } from 'use-context-selector';
import * as React from 'react';
import { MetricsDescendantsProvider, MetricsHeaderDescendantsProvider, MetricsSecondaryDescendantsProvider, useMetricsDescendants, useMetricsFooterDescendants, useMetricsHeaderDescendants, useMetricsSecondaryDescendants, MetricsFooterDescendantsProvider, useMetricsMainDescendants, MetricsMainDescendantsProvider } from './useMetrics';
import { ResizeObserver } from '@universe-design/react-utilities';
import { forwardHTMLRef } from '../internal';
import { useCssScope } from '../config-provider';
import clsx from 'clsx';
import { useSyncExternalStore } from 'use-sync-external-store/shim';
const {
  useContextSelector: useMetricsGroupContextSelector,
  Provider: MetricsGroupContextProvider
} = createContextHook('MetricsGroupContext', {
  strict: false,
  customizedContextMethods: {
    createContext,
    useContext,
    useContextSelector
  },
  defaultValue: null
});
const MetricsGroup = forwardHTMLRef((props, ref) => {
  const {
    titleLineClamp = 2,
    className,
    ...restProps
  } = props;
  const [location, setLocation] = React.useState([]);
  const [headerHeights, setHeaderHeights] = React.useState([]);
  const [mainHeights, setMainHeights] = React.useState([]);
  const [secondaryHeights, setSecondaryHeights] = React.useState([]);
  const [footerHeights, setFooterHeights] = React.useState([]);
  const rootCssScope = useCssScope('metrics');
  const descendants = useMetricsDescendants();
  const {
    list: descendantList
  } = useSyncExternalStore(descendants.subscribe, descendants.getSnapshot, descendants.getSnapshot);
  const headerDescendants = useMetricsHeaderDescendants();
  const {
    list: headerDescendantList
  } = useSyncExternalStore(headerDescendants.subscribe, headerDescendants.getSnapshot, headerDescendants.getSnapshot);
  const mainDescendants = useMetricsMainDescendants();
  const {
    list: mainDescendantList
  } = useSyncExternalStore(mainDescendants.subscribe, mainDescendants.getSnapshot, mainDescendants.getSnapshot);
  const secondaryDescendants = useMetricsSecondaryDescendants();
  const {
    list: secondaryDescendantList
  } = useSyncExternalStore(secondaryDescendants.subscribe, secondaryDescendants.getSnapshot, secondaryDescendants.getSnapshot);
  const footerDescendants = useMetricsFooterDescendants();
  const {
    list: footerDescendantList
  } = useSyncExternalStore(footerDescendants.subscribe, footerDescendants.getSnapshot, footerDescendants.getSnapshot);
  const recalculateLocation = React.useCallback(() => {
    const xy = [];
    for (const d of descendantList) {
      xy.push({
        x: (d === null || d === void 0 ? void 0 : d.node.offsetLeft) || 0,
        y: (d === null || d === void 0 ? void 0 : d.node.offsetTop) || 0
      });
    }
    setLocation(xy);
  }, [descendantList]);
  const recalculateHeaderHeights = React.useCallback(() => {
    const headerHeight = [];
    for (const d of headerDescendantList) {
      headerHeight.push(d === null || d === void 0 ? void 0 : d.node.clientHeight);
    }
    setHeaderHeights(headerHeight);
  }, [headerDescendantList]);
  const recalculateMainHeights = React.useCallback(() => {
    const mainHeight = [];
    for (const d of mainDescendantList) {
      mainHeight.push(d === null || d === void 0 ? void 0 : d.node.clientHeight);
    }
    setMainHeights(mainHeight);
  }, [mainDescendantList]);
  const recalculateSecondaryHeights = React.useCallback(() => {
    const secondaryHeight = [];
    for (const d of secondaryDescendantList) {
      secondaryHeight.push(d === null || d === void 0 ? void 0 : d.node.clientHeight);
    }
    setSecondaryHeights(secondaryHeight);
  }, [secondaryDescendantList]);
  const recalculateFooterHeights = React.useCallback(() => {
    const footerHeight = [];
    for (const d of footerDescendantList) {
      footerHeight.push(d === null || d === void 0 ? void 0 : d.node.clientHeight);
    }
    setFooterHeights(footerHeight);
  }, [footerDescendantList]);
  React.useEffect(() => {
    recalculateLocation();
    recalculateHeaderHeights();
    recalculateMainHeights();
    recalculateSecondaryHeights();
    recalculateFooterHeights();
  }, [recalculateLocation, recalculateHeaderHeights, recalculateMainHeights, recalculateSecondaryHeights, recalculateFooterHeights]);
  const calculateRowInfo = React.useCallback(heights => {
    const rows = [];

    // 按 y 坐标分组
    location.forEach((loc, index) => {
      const existingRow = rows.find(row => row.y === loc.y);
      if (existingRow) {
        existingRow.heights.push(heights[index]);
      } else {
        rows.push({
          y: loc.y,
          heights: [heights[index]]
        });
      }
    });

    // 计算每行的最大高度映射
    const rowMaxHeights = rows.map(row => ({
      y: row.y,
      maxHeight: Math.max(...row.heights)
    }));

    // 为每个位置生成对应的行信息
    return location.map(loc => {
      var _rowMaxHeights$find;
      return {
        row: rows.findIndex(r => r.y === loc.y),
        maxHeight: ((_rowMaxHeights$find = rowMaxHeights.find(r => r.y === loc.y)) === null || _rowMaxHeights$find === void 0 ? void 0 : _rowMaxHeights$find.maxHeight) || 0
      };
    });
  }, [location]);
  const headerRowInfo = React.useMemo(() => calculateRowInfo(headerHeights), [calculateRowInfo, headerHeights]);
  const mainRowInfo = React.useMemo(() => calculateRowInfo(mainHeights), [calculateRowInfo, mainHeights]);
  const secondaryRowInfo = React.useMemo(() => calculateRowInfo(secondaryHeights), [calculateRowInfo, secondaryHeights]);
  const footerRowInfo = React.useMemo(() => calculateRowInfo(footerHeights), [calculateRowInfo, footerHeights]);
  return /*#__PURE__*/React.createElement(MetricsDescendantsProvider, {
    value: descendants
  }, /*#__PURE__*/React.createElement(MetricsHeaderDescendantsProvider, {
    value: headerDescendants
  }, /*#__PURE__*/React.createElement(MetricsSecondaryDescendantsProvider, {
    value: secondaryDescendants
  }, /*#__PURE__*/React.createElement(MetricsFooterDescendantsProvider, {
    value: footerDescendants
  }, /*#__PURE__*/React.createElement(MetricsMainDescendantsProvider, {
    value: mainDescendants
  }, /*#__PURE__*/React.createElement(MetricsGroupContextProvider, {
    value: {
      titleLineClamp,
      recalculateLocation,
      recalculateMainHeights,
      recalculateHeaderHeights,
      recalculateSecondaryHeights,
      recalculateFooterHeights,
      headerRowInfo,
      mainRowInfo,
      secondaryRowInfo,
      footerRowInfo
    }
  }, /*#__PURE__*/React.createElement(ResizeObserver, {
    onResize: recalculateLocation
  }, /*#__PURE__*/React.createElement("div", _extends({
    className: clsx(`${rootCssScope()}__group__wrapper`, className),
    ref: ref
  }, restProps), props.children))))))));
});
export { useMetricsGroupContextSelector, MetricsGroup };
