// @universe-design/react/es/metrics/MetricsHeader.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.u1fssmi4{color:@icon-n2;}
import u1rkqst1 from "./../.stylex/u1rkqst1";
//.u1nvqvck{font-size:calc(1em + 2px);}
import ujhu89j from "./../.stylex/ujhu89j";
//.uqkj31s{vertical-align:calc(0.36 * 1em - 0.5 * (1em + 2px));}
import u1g5q35g from "./../.stylex/u1g5q35g";
//.uly5x3f{width:100%;}
import u13d3rh8 from "./../.stylex/u13d3rh8";
//.u22iagw{display:flex;}
import ukchsq7 from "./../.stylex/ukchsq7";
//.u1guyajx{align-items:baseline;}
import u1gmpn6f from "./../.stylex/u1gmpn6f";
//.u4xv25i{gap:4px;}
import umue7z from "./../.stylex/umue7z";
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx, makeStylexStyles } from '@universe-design/stylex/react';
import InfoOutlined from '@universe-design/icons-react/InfoOutlined';
import stylex from '@universe-design/stylex';
import Tooltip from '../tooltip';
import { Text } from '../text';
import { forwardHTMLRef } from '../util/forwardRef';
import { useMetricsContext } from './MetricsContext';
import { resolveShorthandOld } from '../shared/slot';
import { useMetricsGroupContextSelector } from './MetricsGroup';
import { useMetricsHeaderDescendant } from './useMetrics';
import { useMergedRefs, ResizeObserver } from '@universe-design/react-utilities';
import * as React from 'react';
import clsx from 'clsx';
import './style';
const useMetricsHeaderStylexStyles = makeStylexStyles(_ref => {
  let {
    prefix
  } = _ref;
  return {
    header: {
      "props": {
        "a9b677": u13d3rh8,
        "mc9l5x": ukchsq7,
        "Bt984gj": u1gmpn6f,
        "rmohyg": umue7z
      },
      "alias": `${prefix}__header`
    },
    icon: {
      "props": {
        "sj55zd": u1rkqst1,
        "Be2twd7": ujhu89j,
        "ha4doy": u1g5q35g
      },
      "alias": `${prefix}__icon`
    }
  };
});
export const MetricsHeader = forwardHTMLRef((props, ref) => {
  const {
    mainTitle,
    mainInfoIcon,
    mainHeaderContentAfter,
    className,
    ...restProps
  } = props;
  const {
    rootCssScope,
    size
  } = useMetricsContext();
  const {
    index,
    register
  } = useMetricsHeaderDescendant();
  const recalculateHeaderHeights = useMetricsGroupContextSelector(context => context === null || context === void 0 ? void 0 : context.recalculateHeaderHeights);
  const headerRowMaxHeight = useMetricsGroupContextSelector(context => {
    var _context$headerRowInf, _context$headerRowInf2;
    return context === null || context === void 0 ? void 0 : (_context$headerRowInf = context.headerRowInfo) === null || _context$headerRowInf === void 0 ? void 0 : (_context$headerRowInf2 = _context$headerRowInf[index]) === null || _context$headerRowInf2 === void 0 ? void 0 : _context$headerRowInf2.maxHeight;
  });
  const titleLineClamp = useMetricsGroupContextSelector(context => context === null || context === void 0 ? void 0 : context.titleLineClamp) || 2;
  const styles = useMetricsHeaderStylexStyles({
    prefix: rootCssScope()
  });
  const mainTitleProps = resolveShorthandOld(mainTitle);
  const mainInfoIconProps = resolveShorthandOld(mainInfoIcon);
  const mainHeaderContentAfterProps = resolveShorthandOld(mainHeaderContentAfter);
  const {
    tooltip
  } = mainInfoIconProps || {};
  return jsx(React.Fragment, null, jsx("div", _extends({
    className: clsx(`${rootCssScope()}__header__wrapper`, className),
    style: {
      height: headerRowMaxHeight
    }
  }, restProps), jsx(ResizeObserver, {
    onResize: recalculateHeaderHeights
  }, jsx(Text, {
    as: "div",
    ref: useMergedRefs(ref, register),
    sx: [styles.header],
    variant: size === 'medium' ? 'body0' : 'title5'
  }, mainTitle !== null && jsx(Text, {
    lineClamp: titleLineClamp
  }, mainTitleProps === null || mainTitleProps === void 0 ? void 0 : mainTitleProps.children), mainInfoIcon !== null && jsx(Tooltip, tooltip, jsx("span", null, jsx(InfoOutlined, {
    className: stylex(styles.icon),
    width: 1
  }))), mainHeaderContentAfter !== null && jsx("div", null, mainHeaderContentAfterProps === null || mainHeaderContentAfterProps === void 0 ? void 0 : mainHeaderContentAfterProps.children)))));
});
