// @universe-design/react/es/metrics/MetricsMainContent.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.u1euv43f{position:absolute;}
import u1u1p7mj from "./../.stylex/u1u1p7mj";
//.u1b6n930{transform-origin:left top;}
import uc4lfdq from "./../.stylex/uc4lfdq";
//.uz5stix{white-space:nowrap;}
import uugkpzq from "./../.stylex/uugkpzq";
//.u10pi13n{position:relative;}
import u5jxv71 from "./../.stylex/u5jxv71";
//.u1jo26qt{font-size:36px;}
import u10wmqsv from "./../.stylex/u10wmqsv";
//.ut8aq9l{line-height:42px;}
import u10if3fb from "./../.stylex/u10if3fb";
//.u1rt2boy{font-size:24px;}
import udavndu from "./../.stylex/udavndu";
//.u336tjw{line-height:28px;}
import u903f20 from "./../.stylex/u903f20";
//.uly5x3f{width:100%;}
import u13d3rh8 from "./../.stylex/u13d3rh8";
//.u22iagw{display:flex;}
import ukchsq7 from "./../.stylex/ukchsq7";
//.u122n59{align-items:center;}
import ubyhozq from "./../.stylex/ubyhozq";
//.ulvyvdh{flex-wrap:nowrap;}
import u125g0x8 from "./../.stylex/u125g0x8";
//.u1869bpl{justify-content:space-between;}
import uwg391m from "./../.stylex/uwg391m";
//.ueyvcx7{font-family:Roboto;}
import umbknfa from "./../.stylex/umbknfa";
//.u1z30l8{color:@text-title;}
import u22yt5q from "./../.stylex/u22yt5q";
//.u1hd9hm2{font-weight:500;}
import u1tfkhub from "./../.stylex/u1tfkhub";
/* eslint-disable react/no-unknown-property */
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx, makeStylexStyles } from '@universe-design/stylex/react';
import { ResizeObserver, useMergedRefs } from '@universe-design/react-utilities';
import stylex from '@universe-design/stylex';
import React, { useMemo } from 'react';
import { MetricsChart } from './MetricsChart';
import { forwardHTMLRef } from '../util/forwardRef';
import { useMetricsContext } from './MetricsContext';
import { resolveShorthandOld } from '../shared/slot';
import { useMetricsMainDescendant } from './useMetrics';
import { useMetricsGroupContextSelector } from './MetricsGroup';
const useMetricsMainContentStylexStyles = makeStylexStyles(_ref => {
  let {
    prefix
  } = _ref;
  return {
    root: {
      "props": {
        "a9b677": u13d3rh8,
        "mc9l5x": ukchsq7,
        "Bt984gj": ubyhozq,
        "Eh141a": u125g0x8,
        "Brf1p80": uwg391m,
        "Bahqtrf": umbknfa,
        "sj55zd": u22yt5q,
        "Bhrd7zp": u1tfkhub
      },
      "alias": `${prefix}__root`
    },
    mainWrapper: {
      "props": {
        "mc9l5x": ukchsq7,
        "Bt984gj": ubyhozq
      },
      "alias": `${prefix}__main__wrapper`
    },
    medium: {
      "props": {
        "Be2twd7": udavndu,
        "Bhrd7zp": u1tfkhub,
        "Bg96gwp": u903f20
      },
      "alias": `${prefix}__medium`
    },
    large: {
      "props": {
        "Be2twd7": u10wmqsv,
        "Bg96gwp": u10if3fb
      },
      "alias": `${prefix}__large`
    },
    container: {
      "props": {
        "a9b677": u13d3rh8,
        "qhf8xq": u5jxv71
      },
      "alias": `${prefix}__container`
    },
    content: {
      "props": {
        "qhf8xq": u1u1p7mj,
        "B76gyc8": uc4lfdq,
        "Huce71": uugkpzq
      },
      "alias": `${prefix}__content`
    }
  };
});
export const MetricsMainContent = forwardHTMLRef((props, ref) => {
  const {
    mainValue,
    chart,
    ...restProps
  } = props;
  const {
    rootCssScope,
    size,
    position
  } = useMetricsContext();
  const {
    index,
    register
  } = useMetricsMainDescendant();
  const recalculateMainHeights = useMetricsGroupContextSelector(context => context === null || context === void 0 ? void 0 : context.recalculateMainHeights);
  const mainRowMaxHeight = useMetricsGroupContextSelector(context => {
    var _context$mainRowInfo, _context$mainRowInfo$;
    return context === null || context === void 0 ? void 0 : (_context$mainRowInfo = context.mainRowInfo) === null || _context$mainRowInfo === void 0 ? void 0 : (_context$mainRowInfo$ = _context$mainRowInfo[index]) === null || _context$mainRowInfo$ === void 0 ? void 0 : _context$mainRowInfo$.maxHeight;
  });
  const styles = useMetricsMainContentStylexStyles({
    prefix: rootCssScope()
  });
  const [containerWidth, setContainerWidth] = React.useState(1);
  const [contentWidth, setContentWidth] = React.useState(1);
  const [contentHeight, setContentHeight] = React.useState(1);
  const scale = useMemo(() => {
    if (containerWidth < contentWidth) {
      return containerWidth / contentWidth;
    }
    return 1;
  }, [containerWidth, contentWidth]);
  const mainValueProps = resolveShorthandOld(mainValue);
  const {
    autoScaleContent = true
  } = mainValueProps;
  let content;
  if (!autoScaleContent) {
    content = jsx("span", null, mainValueProps === null || mainValueProps === void 0 ? void 0 : mainValueProps.children);
  } else {
    content = jsx(ResizeObserver, {
      onResize: _ref2 => {
        let {
          width,
          height
        } = _ref2;
        setContentWidth(width);
        setContentHeight(height);
      }
    }, jsx("span", {
      sx: [styles.content],
      style: {
        transform: `scale(${scale})`
      }
    }, mainValueProps === null || mainValueProps === void 0 ? void 0 : mainValueProps.children));
  }
  return jsx("div", _extends({
    sx: [styles.mainWrapper],
    style: {
      height: mainRowMaxHeight
    }
  }, restProps), jsx(ResizeObserver, {
    onResize: recalculateMainHeights
  }, jsx("div", {
    ref: useMergedRefs(ref, register),
    sx: [styles.root, styles[size]]
  }, jsx(ResizeObserver, {
    onResize: _ref3 => {
      let {
        width
      } = _ref3;
      setContainerWidth(width);
    }
  }, jsx("div", {
    sx: [styles.container],
    style: autoScaleContent ? {
      height: contentHeight * scale
    } : {}
  }, content)), position === 'afterMainValue' && jsx(MetricsChart, {
    chart: chart
  }))));
});
