// @universe-design/react/es/tree/hooks/checkStrategy.js
import { CheckedType, ShowCheckedStrategy } from '../interface';
import { isNonLeaf2 } from '../util';
import { getCheckedCache } from './getCheckedCache';

// 尽可能用指针来减少 key 太长带来的内存问题

/**
 * show all
 */
export const checkStrategyShowAll = {
  cache2CheckedSet(doublyTreeNodeMap, cache) {
    const checkedSet = {};
    for (const [k, v] of cache.entries()) {
      checkedSet[k] = v === CheckedType.Checked;
    }
    return checkedSet;
  },
  checkedSet2Cache(doublyTreeNodeMap, checkedSet) {
    const checkedKeys = Object.entries(checkedSet).filter(_ref => {
      let [, v] = _ref;
      return v === true;
    }).map(_ref2 => {
      let [x] = _ref2;
      return x;
    });
    return getCheckedCache({
      doublyTreeNodeMap,
      checkedKeys,
      checkStrictly: false
    });
  }
};

/**
 * show parent
 * 如果所有的 children 都选中则只显示 parent
 */
export const checkStrategyShowParent = {
  ...checkStrategyShowAll,
  cache2CheckedSet(doublyTreeNodeMap, cache) {
    const checkedSet = {};
    // 补充不存在的 value
    for (const [key, checkedType] of cache) {
      if (checkedType === CheckedType.Checked) {
        var _node$parent, _node$parent$children;
        checkedSet[key] = true;
        const node = doublyTreeNodeMap[key];
        const allSiblingsChecked = node === null || node === void 0 ? void 0 : (_node$parent = node.parent) === null || _node$parent === void 0 ? void 0 : (_node$parent$children = _node$parent.children) === null || _node$parent$children === void 0 ? void 0 : _node$parent$children.every(_node => cache.get(_node.__node__.value) === CheckedType.Checked
        //  || isCheckedDisabled(_node.__node__),
        );
        checkedSet[key] = !allSiblingsChecked;
      } else {
        checkedSet[key] = false;
      }
    }
    return checkedSet;
  }
};

/**
 * show children
 * 如果所有的 children 都只显示 parent
 */
export const checkStrategyShowChild = {
  ...checkStrategyShowAll,
  cache2CheckedSet(doublyTreeNodeMap, cache) {
    const checkedSet = {};
    for (const [key, checkedType] of cache.entries()) {
      const node = doublyTreeNodeMap[key];
      // 可能 node 不存在，此时仍然将选择态保留
      const isChild = !node || !isNonLeaf2(node);
      checkedSet[key] = checkedType === CheckedType.Checked && isChild;
    }
    return checkedSet;
  }
};
export const strategyMap = {
  [ShowCheckedStrategy.SHOW_ALL]: checkStrategyShowAll,
  [ShowCheckedStrategy.SHOW_CHILD]: checkStrategyShowChild,
  [ShowCheckedStrategy.SHOW_PARENT]: checkStrategyShowParent
};
