// @universe-design/react/es/tree/hooks/useTree.js
import * as React from 'react';
import { useControlledState } from '@universe-design/react-utilities';
import { treeChildren2TreeNodes, useTreeCssScope } from '../common';
import { ShowCheckedStrategy } from '../interface';
import { TreeContextProvider } from './TreeContext';
import { useTreeSelect } from './useTreeSelect';
import { useTreeCheck } from './useTreeCheck';
import { useTreeExpand } from './useTreeExpand';
import { strategyMap } from './checkStrategy';
import { genDoublyTreeNodeMap } from '../util';
import { useLocale } from '../../config-provider';

/**
 * @docgen
 */

const defaultFilterTreeNode = () => true;
export const useTree = (props, ref) => {
  const locale = useLocale();
  const childrenToTreeData = React.useMemo(() => treeChildren2TreeNodes(props.children), [props.children]);
  const {
    treeData = childrenToTreeData,
    defaultExpandedKeys = [],
    expandedKeys,
    onExpand,
    defaultSelectedKeys = [],
    selectedKeys,
    onSelect,
    filterTreeNode = defaultFilterTreeNode,
    showChecker = false,
    checkable = false,
    selectable = true,
    defaultExpandAll = false,
    multiple,
    showCheckedStrategy = ShowCheckedStrategy.SHOW_ALL,
    switcherIcon,
    labelRenderer,
    loadData,
    onLoad,
    loadedKeys: loadedKeysProp,
    searchValue,
    notFoundContent = locale.global.notFoundContent,
    // virtual scrolling
    virtual = true,
    height,
    nodeRenderer
  } = props;
  const cssScope = useTreeCssScope();
  const [loadedKeys, setLoadedKeys] = useControlledState([], loadedKeysProp);

  // ======== ÕłøÕ╗║õĖĆõĖ¬ÕĖ”ń┤óÕ╝ĢńÜä tree node ÕÅīÕÉæķōŠĶĪ© ========
  const {
    doublyTreeNodeList,
    doublyTreeNodeMap
  } = React.useMemo(() => genDoublyTreeNodeMap(treeData), [treeData]);
  const {
    finalTreeNodeList,
    filteredTreeNodes,
    expandedMap,
    toggleExpanded,
    finalTreeNodeListRef
  } = useTreeExpand({
    doublyTreeNodeMap,
    doublyTreeNodeList,
    defaultExpandAll,
    defaultExpandedKeys,
    expandedKeys,
    onExpand,
    filterTreeNode,
    searchValue
  });
  const treeCheckCtx = useTreeCheck({
    doublyTreeNodeMap,
    checkStrategy: strategyMap[showCheckedStrategy],
    ...(props.checkStrictly === true ? {
      ...props,
      checkStrictly: true
    } : {
      ...props,
      checkStrictly: false
    })
  });
  const treeSelectCtx = useTreeSelect({
    doublyTreeNodeMap,
    defaultSelectedKeys,
    selectedKeys,
    onSelect,
    multiple
  });

  // =========================== List ===========================
  const listRef = React.useRef(null);
  const scrollIntoView = React.useCallback((reference, scrollConfig) => {
    var _doublyTreeNodeMap$ke, _doublyTreeNodeMap$ke2;
    if (!listRef.current) {
      return;
    }
    if (typeof reference === 'number') {
      listRef.current.scrollTo({
        index: reference,
        ...scrollConfig
      });
      return;
    }
    const {
      key
    } = reference;
    const parentKey = (_doublyTreeNodeMap$ke = doublyTreeNodeMap[key]) === null || _doublyTreeNodeMap$ke === void 0 ? void 0 : (_doublyTreeNodeMap$ke2 = _doublyTreeNodeMap$ke.__parent__) === null || _doublyTreeNodeMap$ke2 === void 0 ? void 0 : _doublyTreeNodeMap$ke2.value;
    if (parentKey && !expandedMap[parentKey]) {
      toggleExpanded(parentKey);
    }
    /**
     * Ķ┐Öķćīķ£ĆĶ”üõĖĆõĖ¬Õ╝éµŁźÕżäńÉå’╝Ü
     * ńŁēÕŠģĶŖéńé╣Õ▒ĢÕ╝ĆŃĆüõĖŗõĖĆµ¼Ī render õ╣ŗÕÉÄ’╝īµēŹĶāĮµŗ┐Õł░µø┤µ¢░ÕÉÄńÜä finalTreeNodeList
     * */
    setTimeout(() => {
      const index = finalTreeNodeListRef.current.findIndex(n => n.key === key);
      if (index !== -1) {
        var _listRef$current;
        (_listRef$current = listRef.current) === null || _listRef$current === void 0 ? void 0 : _listRef$current.scrollTo({
          index,
          ...scrollConfig
        });
      }
    });
  }, [expandedMap, listRef, finalTreeNodeListRef, toggleExpanded, doublyTreeNodeMap]);
  React.useImperativeHandle(ref, () => ({
    toggleExpanded,
    scrollIntoView
  }), [scrollIntoView, toggleExpanded]);
  return React.useMemo(() => ({
    loadData,
    onLoad,
    loadedKeys,
    selectable,
    checkable,
    setLoadedKeys,
    rootCssScope: cssScope,
    nodeRenderer,
    showChecker,
    expandedMap,
    toggleExpanded,
    switcherIcon,
    finalTreeNodeList,
    labelRenderer,
    filteredTreeNodes,
    ...treeCheckCtx,
    ...treeSelectCtx,
    height,
    virtual,
    listRef,
    searchValue,
    notFoundContent
  }), [checkable, cssScope, expandedMap, filteredTreeNodes, finalTreeNodeList, labelRenderer, loadData, loadedKeys, nodeRenderer, onLoad, selectable, setLoadedKeys, showChecker, switcherIcon, toggleExpanded, treeCheckCtx, treeSelectCtx, height, virtual, searchValue, notFoundContent]);
};
export const TreeProvider = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    children,
    ...props
  } = _ref;
  const treeCtxValue = useTree(props, ref);
  return /*#__PURE__*/React.createElement(TreeContextProvider, {
    value: treeCtxValue
  }, children);
});
