// @universe-design/react/es/tree/hooks/useTreeCheck.js
import { useControlledState } from '@universe-design/react-utilities';
import React from 'react';
import { genArray, isFunction } from '../../util';
import { CheckedType } from '../interface';
import { array2Set, isCheckedDisabled } from '../util';
import { useTreeCheckedContext } from './TreeCheckedContext';
import { getCheckedCache } from './getCheckedCache';
export function useTreeCheck(props) {
  let {
    skipped: skippedOpt = false
  } = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  const {
    doublyTreeNodeMap,
    checkStrategy,
    checkStrictly,
    defaultCheckedKeys,
    checkedKeys: checkedKeysProp
  } = props;

  // 这段逻辑保证单独使用 useTreeCheck 也一定没有问题
  const treeValueContext = useTreeCheckedContext();
  const skipped = Boolean(treeValueContext) || skippedOpt;

  // eslint-disable-next-line prefer-const
  let [checkedKeys, setCheckedKeys] = useControlledState(defaultCheckedKeys !== null && defaultCheckedKeys !== void 0 ? defaultCheckedKeys : checkStrictly ? {
    halfChecked: [],
    checked: []
  } : [], checkedKeysProp, {
    isControlledOverride: 'checkedKeys' in props
  });
  let checkedCache = React.useMemo(() => {
    if (skipped) {
      return new Map();
    }
    if (checkStrictly) {
      const cache = new Map();
      const {
        checked,
        halfChecked
      } = checkedKeys;
      checked.forEach(k => cache.set(k, CheckedType.Checked));
      halfChecked.forEach(k => cache.set(k, CheckedType.Indeterminate));
      Object.keys(doublyTreeNodeMap).forEach(key => {
        if (cache.get(key) === undefined) {
          cache.set(key, CheckedType.Unchecked);
        }
      });
      return cache;
    } else {
      const keys = genArray(checkedKeys);
      return checkStrategy.checkedSet2Cache(doublyTreeNodeMap, array2Set(keys));
    }
  }, [checkStrategy, checkStrictly, checkedKeys, doublyTreeNodeMap, skipped]);
  const checkedSet = React.useMemo(() => {
    if (skipped) {
      return {};
    }
    if (checkStrictly) {
      const r = {};
      for (const [k, v] of checkedCache.entries()) {
        if (v === CheckedType.Checked) {
          r[k] = true;
        }
      }
      return r;
    }
    return checkStrategy.cache2CheckedSet(doublyTreeNodeMap, checkedCache);
  }, [skipped, checkStrictly, checkStrategy, doublyTreeNodeMap, checkedCache]);
  if (treeValueContext) {
    ({
      checkedCache,
      setCheckedKeys
    } = treeValueContext);
  }

  /**
   * 切换 checked 状态
   */
  const toggleChecked = React.useCallback(key => {
    const node = doublyTreeNodeMap[key];
    if (!node || isCheckedDisabled(node.__node__)) {
      return;
    }
    const checkedType = checkedCache.get(node.__node__.value);
    let keyToRemove;
    let keyToAdd;
    if (checkedType === CheckedType.Checked) {
      keyToRemove = node.key;
    } else {
      keyToAdd = node.key;
    }
    if (props.checkStrictly) {
      const checkedCache = getCheckedCache({
        doublyTreeNodeMap,
        checkedKeys: checkedKeys,
        keyToAdd,
        keyToRemove,
        checkStrictly: true
      });
      const newCheckedKeys = {
        checked: [],
        halfChecked: []
      };
      for (const [k, v] of checkedCache.entries()) {
        if (v === CheckedType.Checked) {
          newCheckedKeys.checked.push(k);
        } else if (v === CheckedType.Indeterminate) {
          newCheckedKeys.halfChecked.push(k);
        }
      }
      setCheckedKeys(newCheckedKeys, () => isFunction(props.onCheck) && props.onCheck(newCheckedKeys, {
        checked: checkedType !== CheckedType.Checked,
        key
      }));
    } else {
      const checkedCache = getCheckedCache({
        doublyTreeNodeMap,
        checkedKeys: checkedKeys,
        keyToAdd,
        keyToRemove,
        checkStrictly: false
      });
      const checkedMap = checkStrategy.cache2CheckedSet(doublyTreeNodeMap, checkedCache);
      const newKeys = Object.keys(checkedMap).filter(k => checkedMap[k]);
      setCheckedKeys(newKeys, () => isFunction(props.onCheck) && props.onCheck(newKeys, {
        checked: checkedType !== CheckedType.Checked,
        key
      }));
    }
  }, [doublyTreeNodeMap, checkedCache, props, checkedKeys, setCheckedKeys, checkStrategy]);
  return React.useMemo(() => ({
    checkedCache,
    checkedSet,
    setCheckedKeys,
    toggleChecked
  }), [checkedCache, checkedSet, setCheckedKeys, toggleChecked]);
}
