// @universe-design/react/lib/date-picker/RangePicker/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RangePickerProvider = exports.RangePicker = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _RangePickerTrigger = require("./RangePickerTrigger");
var _RangePickerPanel = require("./RangePickerPanel");
var _Context = require("./Context");
var _useRangePicker = require("./useRangePicker");
var _RangePickerFooter = require("./RangePickerFooter");
require("../style");
const RangePickerProvider = exports.RangePickerProvider = /*#__PURE__*/_react.default.forwardRef((_ref, ref) => {
  let {
    children,
    ...props
  } = _ref;
  const rangePickerCtx = (0, _useRangePicker.useRangePicker)(props);
  _react.default.useImperativeHandle(ref, () => rangePickerCtx);
  return /*#__PURE__*/_react.default.createElement(_Context.RangePickerContextProvider, {
    value: rangePickerCtx
  }, children);
});
const RangePicker = props => {
  const {
    dateRender,
    startOfWeek,
    hideWeekNumbers,
    janX,
    ranges,
    renderExtraFooter,
    ...restProps
  } = props;
  return /*#__PURE__*/_react.default.createElement(RangePickerProvider, restProps, /*#__PURE__*/_react.default.createElement(_RangePickerTrigger.RangePickerTrigger, (0, _extends2.default)({}, props, {
    panel: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_RangePickerPanel.RangePickerPanel, {
      dateRender: dateRender,
      startOfWeek: startOfWeek,
      hideWeekNumbers: hideWeekNumbers,
      janX: janX
    }), /*#__PURE__*/_react.default.createElement(_RangePickerFooter.RangePickerFooter, {
      ranges: ranges,
      renderExtraFooter: renderExtraFooter
    }))
  })));
};
exports.RangePicker = RangePicker;
