// @universe-design/react/lib/date-picker/RangePicker/RangePickerPanel.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RangePickerPanel = RangePickerPanel;
var _uuf62k = _interopRequireDefault(require("./../../.stylex/uuf62k4"));
var _ux5slyd = _interopRequireDefault(require("./../../.stylex/ux5slyd"));
var _u1ntxetr = _interopRequireDefault(require("./../../.stylex/u1ntxetr"));
var _u1jmgjb = _interopRequireDefault(require("./../../.stylex/u1jmgjb9"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _space = require("../../space");
var _utils = require("../panels/base/utils");
var _Context = require("./Context");
var _RangePickerHeader = _interopRequireDefault(require("./RangePickerHeader"));
var _PanelHeaderSwitchers = require("../panels/PanelHeaderSwitchers");
var _MonthYearSelector = require("../panels/selector/MonthYearSelector");
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _sx = require("../sx");
var _DateCalendarUI = require("../panels/DateCalendar/DateCalendarUI");
require("../style");
//.u18og1rn{margin-top:15px;}
//.u1g777gl{margin-right:15px;}
//.u6jp0wq{margin-bottom:15px;}
//.u1abw64p{margin-left:15px;}
/**
 * 多层选择器解决方案:
 * 当前mode下面，选择日期，才能生效
 * 其他mode上面选择日期，都只是一个中间态，可以在panel组件内部解决
 */
function getChanged(date, mode) {
  const seq = ['date', 'month', 'year'];
  const index = seq.indexOf(mode);
  if (index < 0) {
    return {};
  }
  return seq.slice(index).reduce((pre, mod) => {
    pre[mod] = date.get(mod);
    return pre;
  }, {});
}
function RangePickerPanel(props) {
  // Props
  const {
    className,
    dateRender,
    startOfWeek,
    hideWeekNumbers,
    janX
  } = props;

  // Context
  const {
    prefixCls,
    viewValue,
    mode,
    handleDateSelect,
    handleDateCellMouseEnter,
    handleDateCellMouseLeave,
    isDisabled,
    isPreviewDate,
    localMode,
    setLocalMode,
    startPickerValue,
    endPickerValue,
    setStartPickerValue,
    setEndPickerValue,
    cssScope,
    panelIntervalLimits,
    maxDate,
    minDate
  } = (0, _Context.useRangePickerContext)();
  const panelLimited = panelIntervalLimits === 'adjacent';
  const [focusedIndex, setFocusedIndex] = (0, _react.useState)(0);
  const handleFocusedPanelSelect = date => {
    const newViewDate = date.clone();
    newViewDate.set(getChanged(date, localMode));
    if (focusedIndex === 1) {
      setEndPickerValue(newViewDate);
    } else {
      setStartPickerValue(newViewDate);
    }
    if (mode !== localMode) {
      setLocalMode(mode);
    }
  };
  const handleCollapsed = index => {
    setFocusedIndex(index);
    const nextMode = (0, _utils.getHigherLevelMode)(mode);
    if (nextMode) {
      setLocalMode(nextMode);
    }
  };
  const handleMonthClick = index => {
    setFocusedIndex(index);
    setLocalMode('month');
  };
  const handleYearClick = index => {
    setFocusedIndex(index);
    setLocalMode('year');
  };
  if (localMode !== mode) {
    return /*#__PURE__*/_react.default.createElement(_DateCalendarUI.DateCalendarUI, {
      sx: [{
        "props": {
          "B6of3ja": _uuf62k.default,
          "t21cq0": _ux5slyd.default,
          "jrapky": _u1ntxetr.default,
          "Frg6f3": _u1jmgjb.default
        }
      }],
      key: focusedIndex,
      mode: localMode,
      disabledDate: isDisabled,
      defaultPickerValue: focusedIndex === 0 ? startPickerValue : endPickerValue,
      onSelect: handleFocusedPanelSelect,
      startOfWeek: startOfWeek,
      minDate: minDate,
      maxDate: maxDate,
      hideWeekNumbers: hideWeekNumbers,
      janX: janX
    });
  }
  const getPickerBodyProps = index => {
    const dates = viewValue.filter(x => Boolean(x));
    return {
      disabledDate: isDisabled,
      minDate,
      maxDate,
      previewDate: isPreviewDate,
      pickerValue: index === 0 ? startPickerValue : endPickerValue,
      mode,
      onSelect: handleDateSelect,
      onDateMouseEnter: handleDateCellMouseEnter,
      onContentMouseLeave: handleDateCellMouseLeave,
      startOfWeek,
      hideWeekNumbers,
      janX,
      ...(dateRender ? {
        cellRenderer: (date, _ref) => {
          let {
            viewMode,
            defaultNode
          } = _ref;
          return dateRender(date, viewMode, defaultNode);
        }
      } : {}),
      ...(viewValue[0] && viewValue[1] ? {
        selectedRange: viewValue
      } : {
        selectedDates: dates
      })
    };
  };
  const getMonthYearSelectorProps = index => {
    const MonthYearSelectorCls = cssScope('-panel-header-content__center');
    const resValue = index === 0 ? startPickerValue : endPickerValue;
    return {
      mode,
      pickerValue: resValue,
      className: MonthYearSelectorCls,
      hiddenCollapse: mode === 'year',
      collapsed: localMode === mode,
      onMonthClick: () => handleMonthClick(index),
      onYearClick: () => handleYearClick(index),
      onCollapsed: () => handleCollapsed(index)
    };
  };
  const isDate = mode === 'date';
  const isYear = mode === 'year';

  // TODO handleFns
  const pickerPanelRenderer = index => {
    const isLeft = index === 0;
    const isRight = index === 1;
    const pickerValue = isLeft ? startPickerValue.clone() : endPickerValue.clone();
    const switchPicker = value => {
      if (isLeft) {
        setStartPickerValue(value.clone());
      } else {
        setEndPickerValue(value.clone());
      }
    };

    // handleTimeFns
    const goPrevMonth = () => switchPicker((0, _utils.getPrev)(pickerValue, 'date'));
    const goNextMonth = () => switchPicker((0, _utils.getNext)(pickerValue, 'date'));
    const goPrevYear = () => switchPicker((0, _utils.getPrev)(pickerValue, 'month'));
    const goNextYear = () => switchPicker((0, _utils.getNext)(pickerValue, 'month'));
    const goPre20Years = () => switchPicker((0, _utils.getPrev)(pickerValue, 'year'));
    const goNext20Years = () => switchPicker((0, _utils.getNext)(pickerValue, 'year'));
    return /*#__PURE__*/_react.default.createElement(_DateCalendarUI.DateCalendarUI, (0, _extends2.default)({
      headerRenderer: () => /*#__PURE__*/_react.default.createElement(_RangePickerHeader.default, null, /*#__PURE__*/_react.default.createElement(_space.Space, {
        size: 4,
        className: (0, _stylex.default)(_sx.sxPickerHeaderSwitchWrapper)
      }, isYear && /*#__PURE__*/_react.default.createElement(_PanelHeaderSwitchers.Pre20YearsSwitcher, {
        asPlaceholder: isRight && panelLimited,
        onClick: goPre20Years
      }), !isYear && /*#__PURE__*/_react.default.createElement(_PanelHeaderSwitchers.PreYearSwitcher, {
        asPlaceholder: isRight && panelLimited,
        onClick: goPrevYear
      }), isDate && /*#__PURE__*/_react.default.createElement(_PanelHeaderSwitchers.PreMonthSwitcher, {
        asPlaceholder: isRight && panelLimited,
        onClick: goPrevMonth
      })), /*#__PURE__*/_react.default.createElement(_MonthYearSelector.MonthYearSelector, getMonthYearSelectorProps(index)), /*#__PURE__*/_react.default.createElement(_space.Space, {
        size: 4,
        className: (0, _stylex.default)(_sx.sxPickerHeaderSwitchWrapper)
      }, isDate && /*#__PURE__*/_react.default.createElement(_PanelHeaderSwitchers.NextMonthSwitcher, {
        asPlaceholder: isLeft && panelLimited,
        onClick: goNextMonth
      }), !isYear && /*#__PURE__*/_react.default.createElement(_PanelHeaderSwitchers.NextYearSwitcher, {
        asPlaceholder: isLeft && panelLimited,
        onClick: goNextYear
      }), isYear && /*#__PURE__*/_react.default.createElement(_PanelHeaderSwitchers.Next20YearsSwitcher, {
        asPlaceholder: isLeft && panelLimited,
        onClick: goNext20Years
      })))
    }, getPickerBodyProps(index)));
  };
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(className, `${prefixCls}-panels`)
  }, pickerPanelRenderer(0), pickerPanelRenderer(1)));
}
