// @universe-design/react/lib/date-picker/RangePicker/RangePickerFooter.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RangePickerFooter = void 0;
var _clsx = _interopRequireDefault(require("clsx"));
var _react = _interopRequireDefault(require("react"));
var _button = require("../../button");
var _hooks = require("../../config-provider/hooks");
var _space = require("../../space");
var _util = require("../../util");
var _Context = require("../RangePicker/Context");
require("../style");
const RangePickerFooter = props => {
  const {
    renderExtraFooter,
    ranges
  } = props;
  const pickerFooterCssScope = (0, _hooks.useCssScope)('picker__footer');
  const rangePickerCtx = (0, _Context.useRangePickerContext)();
  const {
    localMode,
    value,
    setValue,
    setOpened
  } = rangePickerCtx;
  const footerContent = [];
  const onQuickSelect = selectedValue => {
    setValue(selectedValue);
    setOpened(false);
  };
  if (renderExtraFooter) {
    footerContent.push(/*#__PURE__*/_react.default.createElement("div", {
      key: "extra",
      className: pickerFooterCssScope('__extra')
    }, renderExtraFooter(rangePickerCtx)));
  } else if (localMode === 'date' && ranges) {
    footerContent.push(/*#__PURE__*/_react.default.createElement("div", {
      key: "default",
      className: pickerFooterCssScope('__range')
    }, /*#__PURE__*/_react.default.createElement(_space.Space, {
      size: "small",
      wrap: true
    }, Object.entries(ranges).map(_ref => {
      var _value$, _value$2;
      let [btnText, rangeValueOrFunc] = _ref;
      let rangeValue;
      if ((0, _util.isFunction)(rangeValueOrFunc)) {
        rangeValue = rangeValueOrFunc();
      } else {
        rangeValue = rangeValueOrFunc;
      }
      const isSelected = (value === null || value === void 0 ? void 0 : (_value$ = value[0]) === null || _value$ === void 0 ? void 0 : _value$.toString()) === rangeValue[0].toString() && (value === null || value === void 0 ? void 0 : (_value$2 = value[1]) === null || _value$2 === void 0 ? void 0 : _value$2.toString()) === rangeValue[1].toString();
      return /*#__PURE__*/_react.default.createElement(_button.Button, {
        className: (0, _clsx.default)(pickerFooterCssScope('__range__preset-btn')),
        key: btnText,
        type: "text",
        selected: isSelected,
        color: "primary",
        onClick: () => onQuickSelect(rangeValue)
      }, btnText);
    })), /*#__PURE__*/_react.default.createElement("div", {
      style: {
        flex: 1
      }
    })));
  }
  if (footerContent.length) {
    return /*#__PURE__*/_react.default.createElement("div", {
      className: pickerFooterCssScope()
    }, footerContent);
  }
  return null;
};
exports.RangePickerFooter = RangePickerFooter;
