// @universe-design/react/src/back-top/index.tsx
import type { FC } from 'react';
import React, { useEffect, useRef, useState } from 'react';
import clsx from 'clsx';
import throttle from 'lodash/throttle';
import { animationScrollTo } from '../util/scrollTo';
import { useCssScope } from '../config-provider/hooks';

import './style';

export interface BackTopProps extends React.PropsWithChildren<unknown> {
  className?: string;
  visibilityHeight?: number;
  target?: () => HTMLElement | typeof window;
  duration?: number;
  onClick?: () => void;
}

export const BackTop: FC<BackTopProps & React.RefAttributes<HTMLDivElement>> = props => {
  const {
    className,
    visibilityHeight = 400,
    children,
    duration = 450,
    target = () => window,
    onClick,
    ...divAttributes
  } = props;

  const cssScope = useCssScope('backtop');
  const eventEl = useRef<HTMLElement | typeof window>();
  const [visible, setVisible] = useState<boolean>(false);

  const classNames = clsx(cssScope(), { [cssScope('--visible')]: visible }, className);

  const getScrollTop = (element: HTMLElement | Window): number => {
    if (element !== null && element !== undefined && element === window) {
      return window.pageYOffset;
    } else if (element instanceof HTMLElement) {
      return element.scrollTop;
    }
    return 0;
  };

  const handleClick = () => {
    if (eventEl.current) {
      animationScrollTo(0, {
        getContainer: () => eventEl.current,
      });
    }
    onClick && onClick();
  };

  const handleScroll = (e: any) => {
    setVisible(getScrollTop(e.target) > visibilityHeight);
  };

  const run = throttle(handleScroll, duration);

  useEffect(() => {
    eventEl.current = target();
    if (eventEl.current) {
      eventEl.current.addEventListener('scroll', run);
    }
    return () => {
      if (eventEl.current) {
        eventEl.current.removeEventListener('scroll', run);
      }
    };
  });

  const renderChildren = () => {
    if (!children) {
      return <div className={cssScope('--default-item')}>top</div>;
    }
    return children;
  };

  return (
    <div className={classNames} {...divAttributes} onClick={handleClick}>
      {renderChildren()}
    </div>
  );
};
