// @byted-service/logagent/lib/protocol/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.quickEncode = exports.decode = exports.encode = void 0;
const tslib_1 = require("tslib");
const google_protobuf_1 = tslib_1.__importDefault(require("google-protobuf"));
const consts_1 = require("../consts");
/**
 * message RequestPayload {
 *   message Tags {
 *     string key = 1;
 *     string value = 2;
 *   }
 *
 *   bytes msg = 1;
 *   repeated Tags tags = 2;
 * }
 */
/**
 * 目前先手写实现，后续再考虑根据情况切换为生成代码
 */
const encode = ({ msg, tags, }) => {
    const writer = new google_protobuf_1.default.BinaryWriter();
    writer.writeBytes(1, msg);
    writer.writeRepeatedMessage(2, tags, (tag, writer) => {
        writer.writeString(1, tag.key);
        writer.writeString(2, tag.value);
    });
    return writer.getResultBuffer();
};
exports.encode = encode;
/**
 * 目前先手写实现，后续再考虑根据情况切换为生成代码
 */
/* istanbul ignore next  */
const decode = (data) => {
    const result = {
        msg: new Uint8Array(),
        tags: [],
    };
    const reader = new google_protobuf_1.default.BinaryReader(data);
    while (reader.nextField()) {
        if (reader.isEndGroup()) {
            break;
        }
        const field = reader.getFieldNumber();
        switch (field) {
            case 1:
                result.msg = reader.readBytes();
                break;
            case 2:
                reader.readMessage(null, (msg, reader) => {
                    const tag = {};
                    while (reader.nextField()) {
                        if (reader.isEndGroup()) {
                            break;
                        }
                        const field = reader.getFieldNumber();
                        switch (field) {
                            case 1:
                                tag.key = reader.readString();
                                break;
                            case 2:
                                tag.value = reader.readString();
                                result.tags.push(tag);
                                break;
                            default:
                                reader.skipField();
                                break;
                        }
                    }
                });
                break;
            default:
                reader.skipField();
                break;
        }
    }
    return result;
};
exports.decode = decode;
const WireType = {
    INVALID: -1,
    VARINT: 0,
    FIXED64: 1,
    DELIMITED: 2,
    START_GROUP: 3,
    END_GROUP: 4,
    FIXED32: 5,
};
const writeInt = (arr, offset, value) => {
    /* istanbul ignore else  */
    if (value < 0x80) {
        // 1 ~ 7
        arr[offset] = value;
        return 1;
    }
    else if (value < 0x4000) {
        // 1 ~ 7
        arr[offset - 1] = (value & 0x7f) | 0x80;
        value = value >>> 7;
        // 8 ~ 14
        arr[offset] = value;
        return 2;
    }
    else if (value < 0x200000) {
        // 1 ~ 7
        arr[offset - 2] = (value & 0x7f) | 0x80;
        value = value >>> 7;
        // 8 ~ 14
        arr[offset - 1] = (value & 0x7f) | 0x80;
        value = value >>> 7;
        // 15 ~ 21
        arr[offset] = value;
        return 3;
    }
    else if (value < 0x10000000) {
        // 1 ~ 7
        arr[offset - 3] = (value & 0x7f) | 0x80;
        value = value >>> 7;
        // 8 ~ 14
        arr[offset - 2] = (value & 0x7f) | 0x80;
        value = value >>> 7;
        // 15 ~ 21
        arr[offset - 1] = (value & 0x7f) | 0x80;
        value = value >>> 7;
        // 22 ~ 28
        arr[offset] = value;
        return 4;
    }
    else {
        const _arr = [];
        while (value > 127) {
            _arr[offset - 3] = (value & 0x7f) | 0x80;
            value = value >>> 7;
        }
        _arr[offset] = value;
        arr.set(_arr, offset);
        return _arr.length;
    }
};
const writeString = (arr, offset, value) => {
    const start = offset;
    let prev = offset;
    for (let i = value.length - 1; i >= 0; i--) {
        let c = value.charCodeAt(i);
        if (c < 128) {
            arr[offset] = c;
            prev = offset;
            offset -= 1;
        }
        else if (c < 2048) {
            arr[offset - 1] = (c >> 6) | 192;
            arr[offset] = (c & 63) | 128;
            prev = offset;
            offset -= 2;
        }
        else if (c < 65536) {
            // Look for surrogates
            if (c >= 0xd800 && c <= 0xdbff && i + 1 < value.length) {
                const second = value.charCodeAt(i + 1);
                if (second >= 0xdc00 && second <= 0xdfff) {
                    offset = prev;
                    // low surrogate
                    // http://mathiasbynens.be/notes/javascript-encoding#surrogate-formulae
                    c = (c - 0xd800) * 0x400 + second - 0xdc00 + 0x10000;
                    arr[offset - 3] = (c >> 18) | 240;
                    arr[offset - 2] = ((c >> 12) & 63) | 128;
                    arr[offset - 1] = ((c >> 6) & 63) | 128;
                    arr[offset] = (c & 63) | 128;
                    prev = offset;
                    offset -= 4;
                }
            }
            else {
                arr[offset - 2] = (c >> 12) | 224;
                arr[offset - 1] = ((c >> 6) & 63) | 128;
                arr[offset] = (c & 63) | 128;
                prev = offset;
                offset -= 3;
            }
        }
    }
    return start - offset;
};
/**
 * logagent 对于 protobuf 的使用比较固定，所以这里用自行实现的倒序写入来提升 encode 性能
 */
const tpl = new Uint8Array(consts_1.maxPacketLen + 1);
const quickEncode = (msg, tags) => {
    let offset = tpl.length - 1;
    for (let i = tags.length - 1; i >= 0; i--) {
        const key = tags[i].key;
        const value = tags[i].value;
        const start = offset;
        const vLen = writeString(tpl, offset, value);
        offset -= vLen;
        offset -= writeInt(tpl, offset, vLen);
        offset -= writeInt(tpl, offset, 2 * 8 + WireType.DELIMITED);
        const kLen = writeString(tpl, offset, key);
        offset -= kLen;
        offset -= writeInt(tpl, offset, kLen);
        offset -= writeInt(tpl, offset, 1 * 8 + WireType.DELIMITED);
        offset -= writeInt(tpl, offset, start - offset);
        offset -= writeInt(tpl, offset, 2 * 8 + WireType.DELIMITED);
    }
    offset -= msg.byteLength;
    // if (offset < 0 + 3 + 1) {
    if (offset < 4) {
        const err = new Error('packet is too long');
        // err[byteLengthSym] = tpl.byteLength - offset - 1 + 3 + 1;
        err[consts_1.byteLengthSym] = tpl.byteLength - offset + 3;
        throw err;
    }
    tpl.set(msg, offset + 1);
    offset -= writeInt(tpl, offset, msg.byteLength);
    offset -= writeInt(tpl, offset, 1 * 8 + WireType.DELIMITED);
    return tpl.slice(offset + 1);
};
exports.quickEncode = quickEncode;
