// @universe-design/react/es/collapse/Collapse.js
import React, { isValidElement, useEffect, useState } from 'react';
import clsx from 'clsx';
import { useCssScope } from '../config-provider/hooks';
import { toFlatArray } from '../util';
import noop from 'lodash/noop';
import './style';
function getActiveKeysArray(activeKey) {
  if (!activeKey) {
    return [];
  }
  let currentActiveKey = activeKey;
  if (!Array.isArray(currentActiveKey)) {
    currentActiveKey = currentActiveKey ? [currentActiveKey] : [];
  }
  return currentActiveKey.map(key => String(key));
}

/**
 * @public
 */

export const Collapse = props => {
  const {
    prefixCls: prefixClsInProps,
    expandIconPosition = 'right',
    className,
    children,
    collapsible,
    accordion,
    activeKey: activeKeyProps,
    defaultActiveKey,
    destroyInactivePanel,
    onChange = noop,
    expandIcon,
    style,
    type = 'default'
  } = props;
  const [activeKey, setActiveKey] = useState(getActiveKeysArray(activeKeyProps !== null && activeKeyProps !== void 0 ? activeKeyProps : defaultActiveKey));
  const cssScope = useCssScope('collapse');
  const prefixCls = prefixClsInProps || cssScope();
  const collapseClassName = clsx(prefixCls, [`${prefixCls}-icon-position-${expandIconPosition}`], className, {
    [`${prefixCls}-obvious-title`]: type === 'obvious'
  });
  useEffect(() => {
    if (activeKeyProps) {
      setActiveKey(getActiveKeysArray(activeKeyProps));
    }
  }, [activeKeyProps]);
  const updateActiveKey = newActiveKey => {
    if (!('activeKey' in props)) {
      setActiveKey(newActiveKey);
    }
    onChange(accordion ? newActiveKey[0] : newActiveKey);
  };
  const handleItemClick = panelKey => {
    let newActiveKey = [];
    if (accordion) {
      newActiveKey = activeKey[0] === panelKey ? [] : [panelKey];
    } else {
      const isActive = activeKey.indexOf(panelKey) > -1;
      if (isActive) {
        newActiveKey = activeKey.filter(key => key !== panelKey);
      } else {
        newActiveKey = activeKey.concat(panelKey);
      }
    }
    updateActiveKey(newActiveKey);
  };
  return /*#__PURE__*/React.createElement("div", {
    style: style,
    className: collapseClassName,
    role: accordion ? 'tablist' : undefined
  }, (() => toFlatArray(children).map((child, index) => {
    if (! /*#__PURE__*/isValidElement(child) || typeof child.type === 'string') {
      return child;
    }
    const {
      collapsible: childCollapsible,
      destroyInactivePanel: childDestroyInactivePanel
    } = child.props;
    const mergeCollapsible = childCollapsible !== null && childCollapsible !== void 0 ? childCollapsible : collapsible;
    const mergeDestroyInactivePanel = childDestroyInactivePanel !== null && childDestroyInactivePanel !== void 0 ? childDestroyInactivePanel : destroyInactivePanel;
    const key = child.key || String(index);
    let isActive = false;
    if (accordion) {
      isActive = activeKey[0] === key;
    } else {
      isActive = activeKey.indexOf(key) > -1;
    }
    const childProps = {
      ...child.props,
      key,
      panelKey: key,
      collapsible: mergeCollapsible,
      isActive,
      onItemClick: handleItemClick,
      destroyInactivePanel: mergeDestroyInactivePanel,
      expandIconPosition,
      prefixCls,
      expandIcon,
      accordion
    };
    return /*#__PURE__*/React.cloneElement(child, childProps);
  }))());
};
