// @universe-design/react/es/collapse/CollapsePanel.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React from 'react';
import clsx from 'clsx';
import noop from 'lodash/noop';
import stylex from '@universe-design/stylex';
import { PanelContent } from './PanelContent';
import DownBoldOutlined from '@universe-design/icons-react/DownBoldOutlined';
import ExpandRightFilled from '@universe-design/icons-react/ExpandRightFilled';
import { CSSTransition } from 'react-transition-group';
import { collapseTransitionProps } from './transtion';
import { IconButton } from '../button';
import { cssTypographyHeadline } from '../stylex';
import './style';
/**
 *
 * @public
 */

function defaultExpandIcon(props, customIcon) {
  const {
    expandIconPosition,
    isActive,
    prefixCls,
    collapsible
  } = props;
  let icon = null;
  let size;
  if (expandIconPosition === 'left') {
    icon = customIcon || /*#__PURE__*/React.createElement(ExpandRightFilled, {
      style: {
        transform: isActive ? 'rotate(90deg)' : ''
      }
    });
    size = 10;
  } else {
    icon = customIcon || /*#__PURE__*/React.createElement(DownBoldOutlined, {
      style: {
        transform: isActive ? 'rotate(180deg)' : ''
      }
    });
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    size = 16;
  }
  return /*#__PURE__*/React.createElement(IconButton, {
    className: clsx(`${prefixCls}-expand-icon`, {
      [`${prefixCls}-expand-icon--noaction`]: collapsible === 'header',
      [`${prefixCls}-expand-icon-unfold`]: isActive
    }),
    icon: icon,
    size: size
  });
}

/**
 * Collapse.Panel 的属性
 * @public
 */

export const CollapsePanel = props => {
  const {
    collapsible,
    children,
    isActive,
    panelKey,
    onItemClick = noop,
    header,
    extra,
    expandIcon,
    showArrow = true,
    forceRender,
    destroyInactivePanel,
    expandIconPosition,
    prefixCls,
    style,
    className,
    accordion
  } = props;
  const disabled = collapsible === 'disabled';
  const actualActive = !disabled && isActive;
  const handleClick = () => {
    if (disabled) {
      return;
    }
    onItemClick(String(panelKey));
  };
  return /*#__PURE__*/React.createElement("div", {
    style: style,
    className: clsx(`${prefixCls}-item`, {
      [`${prefixCls}-item-disabled`]: disabled,
      [`${prefixCls}-item-no-arrow`]: !showArrow,
      [`${prefixCls}-item-active`]: actualActive
    }, className)
  }, /*#__PURE__*/React.createElement("div", {
    className: `${prefixCls}-header`,
    onClick: () => {
      collapsible !== 'header' && handleClick();
    },
    role: accordion ? 'tab' : 'button',
    "aria-expanded": actualActive
  }, (() => {
    if (showArrow) {
      const panelProps = {
        isActive: actualActive,
        collapsible,
        key: panelKey,
        expandIconPosition,
        prefixCls
      };
      const icon = defaultExpandIcon(panelProps, expandIcon && expandIcon(panelProps));
      if (! /*#__PURE__*/React.isValidElement(icon)) {
        return icon;
      }
      return /*#__PURE__*/React.cloneElement(icon, {
        className: clsx(icon.props.className, `${prefixCls}-arrow`)
      });
    }
    return null;
  })(), /*#__PURE__*/React.createElement("span", {
    className: clsx(`${prefixCls}-header-text`, stylex(cssTypographyHeadline)),
    onClick: () => collapsible === 'header' && handleClick()
  }, header), extra && /*#__PURE__*/React.createElement("div", {
    className: `${prefixCls}-extra`
  }, extra)), /*#__PURE__*/React.createElement(CSSTransition, _extends({}, collapseTransitionProps, {
    in: actualActive,
    mountOnEnter: !forceRender,
    unmountOnExit: destroyInactivePanel && !forceRender
  }), state => /*#__PURE__*/React.createElement(PanelContent, {
    isActive: actualActive,
    forceRender: forceRender,
    prefixCls: prefixCls,
    className: clsx({
      [`${prefixCls}-content-hidden`]: state === 'exited' && !actualActive,
      [`${prefixCls}-content-transition`]: state === 'entering' || state === 'exiting'
    }),
    role: accordion ? 'tabpanel' : undefined
  }, children)));
};
