// @universe-design/react/src/popconfirm/stories/homeless.stories.tsx
/** @jsxImportSource @emotion/react */
import { Popconfirm, Button } from '../..';
import '../style';
import type { ComponentProps } from 'react';
import React from 'react';
import './index.less';
import { default as placementDemo } from '../demo/placement';

export default {
  title: 'Components/Popconfirm/临时示例',
  component: Popconfirm,
};

export const BasicUsage = Object.assign((props: ComponentProps<typeof Popconfirm>) => (
  <Popconfirm {...props}>Some text</Popconfirm>
));

export const placements = () => <div style={{ marginTop: 200, marginLeft: 200 }}>{placementDemo()}</div>;

const content = (
  <div>
    <div>Content</div>
    <div>Content</div>
  </div>
);

export const Mode = () => (
  <>
    <Popconfirm content={content}>
      <Button type="text" color="danger" href="#">
        Delete
      </Button>
    </Popconfirm>
    <Popconfirm title="Are you sure to delete this task?" content={content}>
      <Button type="text" color="danger" href="#">
        Delete
      </Button>
    </Popconfirm>
  </>
);

export const Disabled = () => (
  <Popconfirm title="Are you sure to delete this task?" content={content} disabled>
    <Button type="primary">禁用</Button>
  </Popconfirm>
);

export const NoIcon = () => (
  <Popconfirm title="Are you sure to delete this task?" content={content} icon={null}>
    <Button type="primary">无图标</Button>
  </Popconfirm>
);
