// @universe-design/react/src/popconfirm/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C6MaxWidth from '../demo/6.max-width';
import CAsync from '../demo/async';
import CBasic from '../demo/basic';
import CDynamicTrigger from '../demo/dynamic-trigger';
import CIcon from '../demo/icon';
import CLocale from '../demo/locale';
import CPlacement from '../demo/placement';
if (!(C6MaxWidth as Story).storyName) {
  (C6MaxWidth as Story).storyName = '6. 最大宽度';
  (C6MaxWidth as Story).parameters = {
    docs: {
      description: {
        story: '可以设置内容区域的最大宽度。\n\n',
      },
    },
  };
}
if (!(CAsync as Story).storyName) {
  (CAsync as Story).storyName = '5. 异步关闭';
  (CAsync as Story).parameters = {
    docs: {
      description: {
        story: '点击确定后异步关闭气泡确认框，例如提交表单。\n\n',
      },
    },
  };
}
if (!(CBasic as Story).storyName) {
  (CBasic as Story).storyName = '0. 基础用法';
  (CBasic as Story).parameters = {
    docs: {
      description: {
        story: '简单的用法。\n\n',
      },
    },
  };
}
if (!(CDynamicTrigger as Story).storyName) {
  (CDynamicTrigger as Story).storyName = '3. 条件触发';
  (CDynamicTrigger as Story).parameters = {
    docs: {
      description: {
        story: '可以判断是否需要弹出。\n\n',
      },
    },
  };
}
if (!(CIcon as Story).storyName) {
  (CIcon as Story).storyName = '4. 自定义图标';
  (CIcon as Story).parameters = {
    docs: {
      description: {
        story:
          '通过`icon`属性进行设置。组件默认内置了`success`、`warning`、`info`、`error`四种类型，也支持用户传入自定义的`ReactNode`元素。\n\n',
      },
    },
  };
}
if (!(CLocale as Story).storyName) {
  (CLocale as Story).storyName = '1. 定制按钮文字';
  (CLocale as Story).parameters = {
    docs: {
      description: {
        story: '设置`okText`与`cancelText`以自定义按钮文字。\n\n',
      },
    },
  };
}
if (!(CPlacement as Story).storyName) {
  (CPlacement as Story).storyName = '2. 位置';
  (CPlacement as Story).parameters = {
    docs: {
      description: {
        story: '支持 12 个不同的方位。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Popconfirm/官网示例',
  tags: ['autodocs'],
};

export { C6MaxWidth, CAsync, CBasic, CDynamicTrigger, CIcon, CLocale, CPlacement };
