// @universe-design/react/es/pagination/internal/Options.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
//.u1pbl6dw{font-variant-numeric:tabular-nums;}
import u1glo1sh from "./../../.stylex/u1glo1sh";
import React from 'react';
import KEYCODE from './KeyCode';
import stylex from '@universe-design/stylex';
class Options extends React.Component {
  constructor() {
    super(...arguments);
    _defineProperty(this, "state", {
      goInputText: ''
    });
    _defineProperty(this, "buildOptionText", value => `${value} ${this.props.locale.items_per_page}`);
    _defineProperty(this, "changeSize", value => {
      this.props.changeSize(Number(value));
    });
    _defineProperty(this, "handleChange", e => {
      this.setState({
        goInputText: e.target.value
      });
    });
    _defineProperty(this, "handleBlur", e => {
      const {
        goButton,
        quickGo,
        rootPrefixCls
      } = this.props;
      const {
        goInputText
      } = this.state;
      if (goButton || goInputText === '') {
        return;
      }
      this.setState({
        goInputText: ''
      });
      if (e.relatedTarget && (e.relatedTarget.className.indexOf(`${rootPrefixCls}-item-link`) >= 0 || e.relatedTarget.className.indexOf(`${rootPrefixCls}-item`) >= 0)) {
        return;
      }
      quickGo === null || quickGo === void 0 ? void 0 : quickGo(this.getValidValue());
    });
    _defineProperty(this, "go", e => {
      const {
        goInputText
      } = this.state;
      if (goInputText === '') {
        return;
      }
      if ('keyCode' in e && e.keyCode === KEYCODE.ENTER || e.type === 'click') {
        var _this$props$quickGo, _this$props;
        this.setState({
          goInputText: ''
        });
        (_this$props$quickGo = (_this$props = this.props).quickGo) === null || _this$props$quickGo === void 0 ? void 0 : _this$props$quickGo.call(_this$props, this.getValidValue());
      }
    });
  }
  getValidValue() {
    const {
      goInputText
    } = this.state;
    return !goInputText || isNaN(Number(goInputText)) ? undefined : Number(goInputText);
  }
  getPageSizeOptions() {
    const {
      pageSize,
      pageSizeOptions
    } = this.props;
    if (pageSizeOptions.some(option => option.toString() === pageSize.toString())) {
      return pageSizeOptions;
    }
    return pageSizeOptions.concat([pageSize.toString()]).sort((a, b) => {
      const numberA = isNaN(Number(a)) ? 0 : Number(a);
      const numberB = isNaN(Number(b)) ? 0 : Number(b);
      return numberA - numberB;
    });
  }
  render() {
    const {
      pageSize,
      locale,
      rootPrefixCls,
      changeSize,
      quickGo,
      goButton,
      selectComponentClass,
      selectPrefixCls,
      disabled
    } = this.props;
    const {
      goInputText
    } = this.state;
    const prefixCls = `${rootPrefixCls}-options`;
    const Select = selectComponentClass;
    let changeSelect = null;
    let goInput = null;
    let gotoButton = null;
    if (!changeSize && !quickGo) {
      return null;
    }
    const pageSizeOptions = this.getPageSizeOptions();
    if (changeSize && Select) {
      const options = pageSizeOptions.map((opt, i) => /*#__PURE__*/React.createElement(Select.Option, {
        key: i,
        value: opt.toString(),
        className: stylex({
          "props": {
            "B7salou": u1glo1sh
          }
        })
      }, this.buildOptionText(opt)));
      changeSelect = /*#__PURE__*/React.createElement(Select, {
        disabled: disabled,
        prefixCls: selectPrefixCls,
        showSearch: false,
        className: `${prefixCls}-size-changer`,
        optionLabelProp: "children",
        dropdownMatchSelectWidth: false,
        value: (pageSize || pageSizeOptions[0]).toString(),
        onChange: this.changeSize,
        getPopupContainer: triggerNode => triggerNode.parentNode
      }, options);
    }
    if (quickGo) {
      if (goButton) {
        gotoButton = typeof goButton === 'boolean' ? /*#__PURE__*/React.createElement("button", {
          type: "button",
          onClick: this.go,
          onKeyUp: this.go,
          disabled: disabled,
          className: `${prefixCls}-quick-jumper-button`
        }, locale.jump_to_confirm) : /*#__PURE__*/React.createElement("span", {
          onClick: this.go,
          onKeyUp: this.go
        }, goButton);
      }
      goInput = /*#__PURE__*/React.createElement("div", {
        className: `${prefixCls}-quick-jumper`
      }, locale.jump_to, /*#__PURE__*/React.createElement("input", {
        disabled: disabled,
        type: "text",
        value: goInputText,
        onChange: this.handleChange,
        onKeyUp: this.go,
        onBlur: this.handleBlur
      }), locale.page, gotoButton);
    }
    return /*#__PURE__*/React.createElement("li", {
      className: prefixCls
    }, changeSelect, goInput);
  }
}
export default Options;
