// @universe-design/react/es/pagination/internal/Pagination.js
import _extends from "@babel/runtime/helpers/esm/extends";
import * as React from 'react';
import clsx from 'clsx';
import Pager from './Pager';
import Options from './Options';
import LOCALE from './locale/zh_CN';
import { isNonNullable } from '../../util';
import { usePagination } from '../usePagination';
import { PaginationContextProvider } from '../PaginationContext';
const defaultItemRender = (page, type, element) => element;
const defaultPageSizeOptions = ['10', '20', '50', '100'];
const Pagination = props => {
  const {
    // core props
    total: totalProp,
    current: currentProp,
    defaultCurrent,
    pageSize: pageSizeProp,
    defaultPageSize,
    siblingCount,
    boundaryCount,
    disabled,
    onChange,
    onShowSizeChange,
    showEllipses,
    enableSiblingWindowBoundary,
    // ui props
    className,
    style,
    selectPrefixCls = 'rc-select',
    prefixCls = 'rc-pagination',
    hideOnSinglePage = false,
    showQuickJumper = false,
    showTitle = true,
    locale = LOCALE,
    itemRender = defaultItemRender,
    totalBoundaryShowSizeChanger = 50,
    showSizeChanger: showSizeChangerProp,
    showTotal,
    // react node prop
    prevIcon,
    nextIcon,
    jumpPrevIcon,
    jumpNextIcon,
    // select prop
    selectComponentClass,
    pageSizeOptions = defaultPageSizeOptions
  } = props;

  /**
   * ================================================
   * data-attributes & aria-attributes
   * ================================================
   */
  const dataOrAriaAttributeProps = Object.keys(props).reduce((prev, key) => {
    if (key.substr(0, 5) === 'data-' || key.substr(0, 5) === 'aria-' || key === 'role') {
      prev[key] = props[key];
    }
    return prev;
  }, {});
  const usePaginationResult = usePagination({
    total: totalProp,
    current: currentProp,
    defaultCurrent,
    pageSize: pageSizeProp,
    defaultPageSize,
    siblingCount,
    boundaryCount,
    disabled,
    showEllipses,
    onPageChange: onChange,
    onPageSizeChange: onShowSizeChange,
    enableSiblingWindowBoundary
  });
  const {
    total,
    current,
    pageSize,
    pagers,
    prevDisabled,
    nextDisabled,
    prevPage,
    nextPage,
    jumpPrevPage,
    jumpNextPage,
    gotoPage,
    handleGoPrev,
    handleGoNext,
    handleJumpPrev,
    handleJumpNext,
    handlePageSizeChange
  } = usePaginationResult;
  const paginationNodeRef = React.useRef(null);
  const showSizeChanger = (() => {
    if (typeof showSizeChangerProp !== 'undefined') {
      return showSizeChangerProp;
    }
    return total > totalBoundaryShowSizeChanger;
  })();
  const changeSize = React.useMemo(() => {
    if (!showSizeChanger) {
      return;
    }
    return handlePageSizeChange;
  }, [handlePageSizeChange, showSizeChanger]);
  const quickGo = React.useMemo(() => {
    if (total <= pageSize) {
      return undefined;
    }
    if (showQuickJumper) {
      return page => {
        if (!isNonNullable(page)) {
          return;
        }
        gotoPage(page);
      };
    }
  }, [gotoPage, pageSize, showQuickJumper, total]);

  // When hideOnSinglePage is true and there is only 1 page, hide the pager
  if (hideOnSinglePage === true && total <= pageSize) {
    return null;
  }

  /**
   * ================================================
   * Total Node
   * ================================================
   */
  const totalText = (() => {
    if (showTotal) {
      return /*#__PURE__*/React.createElement("li", {
        className: `${prefixCls}-total-text`
      }, showTotal(total, [total === 0 ? 0 : (current - 1) * pageSize + 1, current * pageSize > total ? total : current * pageSize]));
    }
    return null;
  })();

  /**
   * ================================================
   * Pager List Node
   * ================================================
   */
  const pagerList = (() => {
    const getItemIcon = (icon, label) => icon || /*#__PURE__*/React.createElement("button", {
      type: "button",
      "aria-label": label,
      className: `${prefixCls}-item-link`
    });
    const nodes = [];
    for (const pager of pagers) {
      if (pager.type === 'prev') {
        nodes.push(/*#__PURE__*/React.createElement("li", {
          key: "prev",
          title: showTitle ? locale.prev_page : undefined,
          onClick: handleGoPrev,
          tabIndex: prevDisabled ? undefined : 0,
          className: clsx(`${prefixCls}-prev`, {
            [`${prefixCls}-disabled`]: prevDisabled
          }),
          "aria-disabled": prevDisabled
        }, (() => {
          const prevButton = itemRender(prevPage, 'prev', getItemIcon(prevIcon, 'prev page'));
          return /*#__PURE__*/React.isValidElement(prevButton) ? /*#__PURE__*/React.cloneElement(prevButton, {
            disabled: prevDisabled
          }) : prevButton;
        })()));
        continue;
      }
      if (pager.type === 'next') {
        nodes.push(/*#__PURE__*/React.createElement("li", {
          key: "next",
          title: showTitle ? locale.next_page : undefined,
          onClick: handleGoNext,
          tabIndex: nextDisabled ? undefined : 0,
          className: clsx(`${prefixCls}-next`, {
            [`${prefixCls}-disabled`]: nextDisabled
          }),
          "aria-disabled": nextDisabled
        }, (() => {
          const nextButton = itemRender(nextPage, 'next', getItemIcon(nextIcon, 'next page'));
          return /*#__PURE__*/React.isValidElement(nextButton) ? /*#__PURE__*/React.cloneElement(nextButton, {
            disabled: nextDisabled
          }) : nextButton;
        })()));
        continue;
      }
      if (pager.type === 'page') {
        const active = current === pager.page;
        nodes.push(/*#__PURE__*/React.createElement(Pager, {
          rootPrefixCls: prefixCls,
          onClick: gotoPage,
          key: pager.page,
          page: pager.page,
          active: active,
          showTitle: showTitle,
          itemRender: itemRender
        }));
        continue;
      }
      if (pager.type === 'jump-prev') {
        // TODO 这里现在 locale text 不对，因为提供了 siblingCount 以后就不只是 3 和 5 了，可能是任意奇数
        // const prevItemTitle = showLessItems ? locale.prev_3 : locale.prev_5;
        nodes.push(/*#__PURE__*/React.createElement("li", {
          key: "jump-prev",
          onClick: handleJumpPrev,
          tabIndex: 0,
          className: clsx(`${prefixCls}-jump-prev`, {
            [`${prefixCls}-jump-prev-custom-icon`]: Boolean(jumpPrevIcon)
          })
        }, itemRender(jumpPrevPage, 'jump-prev', getItemIcon(jumpPrevIcon, 'prev page'))));
        continue;
      }
      if (pager.type === 'jump-next') {
        // const nextItemTitle = showLessItems ? locale.next_3 : locale.next_5;
        nodes.push(/*#__PURE__*/React.createElement("li", {
          key: "jump-next",
          tabIndex: 0,
          onClick: handleJumpNext,
          className: clsx(`${prefixCls}-jump-next`, {
            [`${prefixCls}-jump-next-custom-icon`]: Boolean(jumpNextIcon)
          })
        }, itemRender(jumpNextPage, 'jump-next', getItemIcon(jumpNextIcon, 'next page'))));
        continue;
      }
    }
    return nodes;
  })();

  /**
   * Options
   */
  const goButton = typeof showQuickJumper === 'object' && showQuickJumper.goButton;
  return /*#__PURE__*/React.createElement(PaginationContextProvider, {
    value: usePaginationResult
  }, /*#__PURE__*/React.createElement("ul", _extends({
    className: clsx(prefixCls, className, {
      [`${prefixCls}-disabled`]: disabled
    }),
    style: style
    // TODO 这里类型过不了
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    ,
    unselectable: "unselectable",
    ref: paginationNodeRef
  }, dataOrAriaAttributeProps), totalText, pagerList, /*#__PURE__*/React.createElement(Options, {
    disabled: disabled,
    locale: locale,
    rootPrefixCls: prefixCls,
    selectComponentClass: selectComponentClass,
    selectPrefixCls: selectPrefixCls,
    changeSize: changeSize,
    pageSize: pageSize,
    pageSizeOptions: pageSizeOptions,
    quickGo: quickGo,
    goButton: goButton
  })));
};
export default Pagination;
