// @universe-design/react/src/react-mini-form/stories/formList.stories.tsx
import * as React from 'react';
import type { Path, FormContextType } from '..';
import { Field, Form, useFieldContext, useFormList, useFormContextSelector, FormSelectors, FormList } from '..';

export default {
  title: 'Components/ReactMiniForm/FormList示例', // 注意：这里需要修改成自己对应 Component name。
  component: Form,
};

const DebugValues = () => (
  <FormSelectors
    selectors={{
      values: ctx => ctx.values,
      touched: ctx => ctx.touched,
      errors: ctx => ctx.errors,
    }}
  >
    {state => <pre>{JSON.stringify(state, undefined, 2)}</pre>}
  </FormSelectors>
);

function getUid() {
  return Math.random().toString(36).substr(2, 5);
}

class DataItem {
  uid: string;
  name?: string;
  value?: string;
  constructor(name?: string, value?: string) {
    this.name = name;
    this.value = value;
    this.uid = getUid();
  }
}

const FieldError = ({ name, error: errorProp }: { name?: Path; error?: string }) => {
  const { error } = useFieldContext({ name });
  const errorWillShow = error || errorProp;
  return (
    <div
      style={{
        color: errorWillShow ? 'var(--R500)' : 'inherit',
        transition: 'color 1s',
      }}
    >
      {errorWillShow}
    </div>
  );
};

const BasicFormHook: React.FC = () => {
  const { fields, actions, error } = useFormList({
    name: 'list',
    rowKey: 'uid',
    validate: [
      {
        validate(value) {
          if (value?.length) {
            return '';
          }
          return '至少输入一行';
        },
      },
    ],
  });

  const validateFields = useFormContextSelector(ctx => ctx.validateFields);

  return (
    <div style={{ display: 'grid', gap: 16, justifyContent: 'start', padding: 16 }}>
      <div style={{ display: 'flex' }}>
        整体报错：
        <FieldError error={error} />
      </div>

      {fields.map((field, index) => (
        <div
          key={field.key}
          style={{
            display: 'grid',
            gridAutoFlow: 'column',
            alignItems: 'end',
            justifyContent: 'start',
            gap: 8,
          }}
        >
          <div>
            <pre>label: {JSON.stringify(field.name, undefined)}</pre>
            <Field name={field.name.concat('name')} validate={v => (!v ? `error-${field.name}-name` : undefined)}>
              <input placeholder="xx" />
            </Field>
            <Field name={field.name.concat('value')} validate={v => (!v ? `error-${field.name}-value` : undefined)}>
              <input placeholder="xx" />
            </Field>
          </div>
          <button onClick={() => actions.remove(index)} type="button">
            delete
          </button>
          <button onClick={() => validateFields([field.name.concat('name'), field.name.concat('value')])} type="button">
            validate
          </button>
          <FieldError name={field.name.concat('name')} />
          <FieldError name={field.name.concat('value')} />
        </div>
      ))}

      <button type="button" onClick={() => validateFields()}>
        validate
      </button>
      <button type="button" onClick={() => actions.push(new DataItem('push item'))}>
        push
      </button>
      <button type="button" onClick={() => actions.push(new DataItem('push1'), new DataItem('push2'))}>
        push + 2
      </button>
      <button type="button" onClick={() => actions.unshift(new DataItem('unshift'))}>
        unshift
      </button>
      <button type="button" onClick={() => actions.insert(1, new DataItem('insert'))}>
        在第二行插入一个
      </button>

      <button type="button" onClick={() => actions.move(1, 2)}>
        移动第二行到第三行
      </button>

      <button type="button" onClick={() => actions.swap(0, 1)}>
        交换第一和第二行
      </button>

      <button
        type="button"
        onClick={() => {
          actions.push(new DataItem('末尾行1'), new DataItem('末尾行2'));
          actions.unshift(new DataItem('首行'));
          actions.swap(1, 2);
          actions.insert(2, new DataItem('插入行'));
        }}
      >
        复合操作: (末尾增加两行，头部增加一行，交换二三行，然后在三行插入一行)
      </button>
    </div>
  );
};

export const 基础用法Hook = () => {
  const initData = {
    // 也可以是这种数据结构：['name1', 'name2']
    // 但是这种数据结构就没办法使用 uid 来复用行，某一些情况下可能有问题
    list: [new DataItem('defaultName', 'defaultValue')],
  };

  return (
    <Form initialValues={initData} onSubmit={console.log}>
      <BasicFormHook />
      <DebugValues />
    </Form>
  );
};

export const 基础用法组件 = () => {
  const initData = {
    // 也可以是这种数据结构：['name1', 'name2']
    // 但是这种数据结构就没办法使用 uid 来复用行，某一些情况下可能有问题
    list: [new DataItem('defaultName', 'defaultValue')],
  };

  const form = React.useRef<FormContextType<typeof initData>>(null);

  return (
    <Form
      ref={form}
      initialValues={initData}
      onSubmit={values => {
        console.log('[onSubmit]', values);
      }}
      onValuesChange={values => {
        console.log('[onValuesChange]', values);
      }}
    >
      <FormList name="list" rowKey="uid" validate={value => (value?.length ? '' : '至少输入一行')}>
        {({ fields, actions, error }) => (
          <div
            style={{
              display: 'grid',
              gap: 16,
              justifyContent: 'start',
              padding: 16,
            }}
          >
            <div style={{ display: 'flex' }}>
              整体报错：
              <FieldError error={error} />
            </div>

            {fields.map((field, index) => (
              <div
                key={field.key}
                style={{
                  display: 'grid',
                  gridAutoFlow: 'column',
                  alignItems: 'end',
                  justifyContent: 'start',
                  gap: 8,
                }}
              >
                <div>
                  <pre>label: {JSON.stringify(field.name, undefined)}</pre>
                  <Field name={field.name.concat('name')} validate={v => (!v ? `error-${field.name}-name` : undefined)}>
                    <input placeholder="xx" />
                  </Field>
                  <Field
                    name={field.name.concat('value')}
                    validate={v => (!v ? `error-${field.name}-value` : undefined)}
                  >
                    <input placeholder="xx" />
                  </Field>
                </div>
                <button onClick={() => actions.remove(index)} type="button">
                  delete
                </button>
                <button
                  onClick={() => form.current?.validateFields([field.name.concat('name'), field.name.concat('value')])}
                  type="button"
                >
                  validate
                </button>
                <FieldError name={field.name.concat('name')} />
                <FieldError name={field.name.concat('value')} />
              </div>
            ))}
            <button type="button" onClick={() => form.current?.validateFields()}>
              validate
            </button>
            <button type="button" onClick={() => actions.push(new DataItem('push item'))}>
              push
            </button>
            <button type="button" onClick={() => actions.push(new DataItem('push1'), new DataItem('push2'))}>
              push + 2
            </button>
            <button type="button" onClick={() => actions.unshift(new DataItem('unshift'))}>
              unshift
            </button>
            <button type="button" onClick={() => actions.insert(1, new DataItem('insert'))}>
              在第二行插入一个
            </button>

            <button type="button" onClick={() => actions.move(1, 2)}>
              移动第二行到第三行
            </button>

            <button type="button" onClick={() => actions.swap(0, 1)}>
              交换第一和第二行
            </button>

            <button
              type="button"
              onClick={() => {
                actions.push(new DataItem('末尾行1'), new DataItem('末尾行2'));
                actions.unshift(new DataItem('首行'));
                actions.swap(1, 2);
                actions.insert(2, new DataItem('插入行'));
              }}
            >
              复合操作: (末尾增加两行，头部增加一行，交换二三行，然后在三行插入一行)
            </button>

            <button
              type="button"
              onClick={() => {
                form.current?.setFieldValue('list', []);
              }}
            >
              清空
            </button>
          </div>
        )}
      </FormList>
      <DebugValues />
    </Form>
  );
};
