// @universe-design/styles/cjs/data-v/index.dark.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.datav_line_default = exports.datav_line_axis = exports.datav_function_text_warning = exports.datav_function_text_success = exports.datav_function_text_info = exports.datav_function_text_danger = exports.datav_function_element_warning = exports.datav_function_element_success = exports.datav_function_element_neutral = exports.datav_function_element_info = exports.datav_function_element_danger = exports.datav_fill_hover = exports.datav_categorical_14 = exports.datav_categorical_13 = exports.datav_categorical_12 = exports.datav_categorical_11 = exports.datav_categorical_10 = exports.datav_categorical_09 = exports.datav_categorical_08 = exports.datav_categorical_07 = exports.datav_categorical_06 = exports.datav_categorical_05 = exports.datav_categorical_04 = exports.datav_categorical_03 = exports.datav_categorical_02 = exports.datav_categorical_01 = void 0;
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
exports.datav_categorical_01 = '#3370eb';
exports.datav_categorical_02 = '#1bcebf';
exports.datav_categorical_03 = '#ffc60a';
exports.datav_categorical_04 = '#ed6d0c';
exports.datav_categorical_05 = '#dca1e4';
exports.datav_categorical_06 = '#25b2e5';
exports.datav_categorical_07 = '#6dcdeb';
exports.datav_categorical_08 = '#288fcb';
exports.datav_categorical_09 = '#94b5f5';
exports.datav_categorical_10 = '#8f61d1';
exports.datav_categorical_11 = '#bf78e9';
exports.datav_categorical_12 = '#008280';
exports.datav_categorical_13 = '#27ad8e';
exports.datav_categorical_14 = '#7bc335';
exports.datav_fill_hover = 'rgba(235, 235, 235, 0.05)';
exports.datav_function_element_danger = '#ff615d';
exports.datav_function_element_info = '#5083fb';
exports.datav_function_element_neutral = '#b9bec6';
exports.datav_function_element_success = '#47cfb8';
exports.datav_function_element_warning = '#ff814a';
exports.datav_function_text_danger = '#cd4841';
exports.datav_function_text_info = '#2c5cc5';
exports.datav_function_text_success = '#2d8639';
exports.datav_function_text_warning = '#c75f1c';
exports.datav_line_axis = 'rgba(235, 235, 235, 0.2)';
exports.datav_line_default = 'rgba(235, 235, 235, 0.1)';
