// @universe-design/react/src/button/ButtonBase.tsx
import type { ReactNode } from 'react';
import React, { forwardRef, useCallback, useMemo, useRef } from 'react';
import clsx from 'clsx';
import isNil from 'lodash/isNil';
import { SpinLoadingSvg } from '../_assets/svgs';
import { useComponentSize, useCssScope } from '../config-provider/hooks';
import type { DefaultSizeType, DefaultExtraSizeType } from '../util';
import { getSizeAbbr, deprecatedWarning, warning } from '../util';
import composeRefs from '../util/composeRefs';
import { useLoading } from '../loading/useLoading';

import './style';
import { markAsDisabledableComp } from '../shared/markAsDisabledableComp';

// type ReactButtonProps = React.DetailedHTMLProps<
//   React.ButtonHTMLAttributes<HTMLButtonElement>,
//   HTMLButtonElement
// >;

/**
 * Deprecated 组件的分类：https://bytedance.feishu.cn/docs/doccngkxeLeLStbebeToeKkPW7b#
 * Breaking change
 * - 重写 type，type 表示按钮类型，每种类型有自己的限制条件，如:
 *  - filled 限制背景改变，color 不变
 *  - outlined 限制背不变
 *  - ghost 限制背景透明
 * - 重写样式，降低样式的层叠优先级
 * - 修正 ButtonLink, LinkButton 的实现
 * - 去掉部分 Button 负 margin
 * - 去掉 info danger 属性，统一使用 color
 *  - info => color="primary"
 *  - danger => color="danger"
 * - 去掉 ghost 属性，ghost => type="ghost"
 *
 */

/**
 * @public
 */
export interface BaseButtonProps {
  // 可以去掉组件的所有样式，「这个的需求是什么？」
  prefixCls?: string;
  /**
   * 设置按钮类型
   * @defaultValue `outlined`
   */
  type?: // variant
  | 'filled'
    | 'primary' // legacy
    | 'outlined'
    | 'default' // legacy
    | 'text'
    | 'text-secondary'
    // | 'ghost'
    | 'icon'
    | 'link';
  /**
   * 设置按钮形状
   */
  shape?: 'circle' | 'square' | 'round';
  /**
   * 设置按钮大小
   * @defaultValue `medium`
   */
  size?: DefaultSizeType | DefaultExtraSizeType; // 悬浮按钮需要 extra-large 48px
  /**
   * 设置按钮颜色
   */
  color?: 'primary' | 'danger';
  /**
   * 将按钮宽度调整为其父宽度的选项
   * @defaultValue false
   */
  block?: boolean;
  /**
   * 设置按钮失效状态
   * @defaultValue false
   */
  disabled?: boolean;
  /**
   * 设置按钮载入状态
   * @defaultValue false
   */
  loading?: boolean;
  /**
   * 延迟显示加载状态
   */
  delay?: number;
  /**
   * 图标按钮选中模式
   * @defaultValue false
   */
  selected?: boolean;
  /**
   * 设置按钮的图标组件
   */
  icon?: ReactNode;
  endIcon?: ReactNode;
  noMinWidth?: boolean;
  children?: React.ReactNode;
  // deprecated legacy
  danger?: boolean;
  // deprecated legacy
  info?: boolean;
  /**
   * 选中后是否自动失焦
   * @defaultValue false
   * @deprecated
   */
  autoBlur?: boolean;
}

export type ButtonType = BaseButtonProps['type'];

export interface AnchorButtonProps
  extends Omit<React.AnchorHTMLAttributes<HTMLAnchorElement>, 'type' | 'onClick' | 'color'>,
    BaseButtonProps {
  /**
   * 点击跳转的地址，指定此属性按钮的行为和 a 链接一致
   */
  href: string;
  /**
   * 相当于 a 链接的`target`属性，`href`存在时生效
   */
  target?: string;
  /**
   * 点击按钮时的回调
   */
  onClick?: React.MouseEventHandler<HTMLElement>;
}

export type ButtonSize = NonNullable<BaseButtonProps['size']>;

export interface NativeButtonProps
  extends Omit<React.ButtonHTMLAttributes<HTMLButtonElement>, 'type' | 'onClick' | 'color'>,
    BaseButtonProps {
  /**
   * 设置 `button` 标签的 `type` 值，可选值请参考 [HTML 标准](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#attr-type)
   * @defaultValue `button`
   */
  htmlType: React.ButtonHTMLAttributes<HTMLButtonElement>['type'];
  onClick?: React.MouseEventHandler<HTMLElement>;
}

/**
 * 支持原生`button`的所有事件和属性。
 * @public
 * @docgen
 */
export type ButtonProps = Partial<AnchorButtonProps & NativeButtonProps>;

export type ButtonRefType = HTMLButtonElement | HTMLAnchorElement | undefined;

const customizeSizeSet = new Set<BaseButtonProps['type']>(['icon']);
function getSizeClass(type: BaseButtonProps['type']) {
  return customizeSizeSet.has(type) ? `${type}-size` : 'size';
}

const ButtonCmp = forwardRef<ButtonRefType, ButtonProps>((props, ref) => {
  const {
    prefixCls: customizePrefixCls,
    className,
    children,
    shape,
    htmlType = 'button',
    type = 'outlined',
    size: customizeSize,
    color = 'default',
    selected = false,
    loading: loadingProp = false,
    delay,
    icon,
    endIcon,
    block = false,
    noMinWidth = false,
    onClick,
    danger,
    info,
    ...rest
  } = props;

  warning(isNil(props.autoBlur), deprecatedWarning('button', 'autoBlur'));

  const cssScope = useCssScope('button');
  const prefixCls = customizePrefixCls || cssScope();
  const [loading] = useLoading({ loading: loadingProp, delay });
  const size = useComponentSize();
  const buttonRef = useRef<ButtonRefType>();

  const mergedRef = useMemo(() => composeRefs(ref, buttonRef), [ref, buttonRef]);

  const defaultSizeGlobale = size || 'medium';
  const sizeValue = getSizeAbbr(customizeSize || defaultSizeGlobale);

  let variant = type;
  if (type === 'primary') {
    variant = 'filled';
  }
  if (type === 'default') {
    variant = 'outlined';
  }

  let usedColor = color;
  if (usedColor === 'default' && danger) {
    usedColor = 'danger';
  }
  if (usedColor === 'default' && info) {
    usedColor = 'primary';
  }

  const classes = clsx(
    prefixCls,
    {
      [cssScope(`--${variant}`)]: variant,
      [cssScope(`--${variant}-${usedColor}`)]: usedColor,
      [cssScope(`--${variant}-${usedColor}-loading`)]: loading,
      [cssScope(`--${variant}-${usedColor}-selected`)]: selected,
      [cssScope('--loading')]: loading,
      [cssScope('--selected')]: selected,
      [cssScope(`--${shape}`)]: shape,
      [cssScope(`--${getSizeClass(variant)}-${sizeValue}`)]: sizeValue,
      [cssScope('--disabled')]: rest.disabled,
      [cssScope('--block')]: block,
      [cssScope('--autowidth')]: noMinWidth,
    },
    className,
  );

  let leftIcon = icon;
  let rightIcon = endIcon;
  if (leftIcon && children) {
    leftIcon = <span className={clsx(cssScope('__icon-inline'), cssScope('__icon-inline-start'))}>{leftIcon}</span>;
  }
  if (rightIcon && children) {
    rightIcon = <span className={clsx(cssScope('__icon-inline'), cssScope('__icon-inline-end'))}>{rightIcon}</span>;
  }

  if (loading && !rest.disabled) {
    leftIcon = (
      <span
        className={clsx(
          cssScope('__loading-icon'),
          cssScope('__icon-inline'),
          children && cssScope('__icon-inline-start'),
        )}
      >
        <SpinLoadingSvg />
      </span>
    );
  }

  const handleClick = useCallback(
    (e: React.MouseEvent<HTMLButtonElement | HTMLAnchorElement, MouseEvent>) => {
      if (loading || rest.disabled) {
        e.stopPropagation();
        e.preventDefault();
      } else {
        onClick && onClick(e);
      }
    },
    [loading, rest.disabled, onClick],
  );

  if (rest.href !== undefined) {
    return (
      <a
        data-testid={cssScope()}
        {...rest}
        ref={mergedRef as React.RefCallback<HTMLAnchorElement>}
        onClick={handleClick}
        className={classes}
        // disabled={disabled}
      >
        {leftIcon}
        {children}
        {rightIcon}
      </a>
    );
  }

  return (
    /* eslint-disable react/button-has-type */
    <button
      data-testid={cssScope()}
      {...rest}
      type={htmlType}
      ref={mergedRef as React.RefCallback<HTMLButtonElement>}
      onClick={handleClick}
      className={classes}
    >
      {leftIcon}
      {children}
      {rightIcon}
    </button>
    /* eslint-enable react/button-has-type */
  );
});

export type ButtonCmpType = typeof ButtonCmp & { __UD_BUTTON?: boolean };

const ButtonBase = /* #__PURE__*/ markAsDisabledableComp(ButtonCmp as ButtonCmpType);

ButtonBase.__UD_BUTTON = true;
if (__DEV__) {
  ButtonBase.displayName = 'ButtonBase';
}

export { ButtonBase };
