// @universe-design/react/src/button/Link.tsx
import type { ReactNode } from 'react';
import React, { forwardRef, useImperativeHandle, useRef, useCallback } from 'react';
import clsx from 'clsx';
import { SpinLoadingSvg } from '../_assets/svgs';
import { useCssScope } from '../config-provider/hooks';
import { markAsDisabledableComp } from '../shared/markAsDisabledableComp';

type ReactAnchorProps = React.DetailedHTMLProps<React.AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>;

/**
 * Deprecated
 * 这是一个根据 Universse Design 设计，做的一个 Button 抽象，它强大而灵活
 */
export interface LinkProps extends ReactAnchorProps {
  prefixCls?: string;
  loading?: boolean;
  disabled?: boolean;
  icon?: ReactNode;
}

const LinkCmp = forwardRef<HTMLAnchorElement, LinkProps>((props, ref) => {
  const {
    prefixCls: customizePrefixCls,
    className,
    children,
    loading = false,
    disabled = false,
    icon: iconProp,
    onClick,
    ...rest
  } = props;

  const cssScope = useCssScope('link');
  const prefixCls = customizePrefixCls || cssScope();
  const aRef = useRef<HTMLAnchorElement>();

  useImperativeHandle<HTMLAnchorElement | undefined, HTMLAnchorElement | undefined>(ref, () => aRef.current);

  const classes = clsx(
    prefixCls,
    {
      [cssScope('--loading')]: loading && !disabled,
      [cssScope('--disabled')]: disabled,
    },
    className,
  );

  let icon = iconProp;
  if (loading && !disabled) {
    icon = <SpinLoadingSvg />;
  }

  const handleClick = useCallback(
    (e: React.MouseEvent<HTMLAnchorElement>) => {
      if (loading || disabled) {
        e.stopPropagation();
        e.preventDefault();
      } else {
        onClick && onClick(e);
      }
    },
    [loading, disabled, onClick],
  );

  return (
    <a
      {...rest}
      ref={a => {
        if (a) {
          aRef.current = a;
        }
      }}
      className={classes}
      onClick={handleClick}
      data-testid={cssScope()}
    >
      {icon}
      {children}
    </a>
  );
});

export type LinkType = typeof LinkCmp & {
  __UD_LINK?: boolean;
  __UD_BUTTON?: boolean;
};

const Link = /* #__PURE__*/ markAsDisabledableComp(LinkCmp as LinkType);

Link.__UD_LINK = true;
Link.__UD_BUTTON = true;
if (__DEV__) {
  Link.displayName = 'Link';
}

export { Link };
