// @universe-design/react/lib/table/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  Table: true
};
exports.Table = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _rcTable = _interopRequireWildcard(require("@universe-design/rc-table"));
var _Table = require("@universe-design/rc-table/lib/Table");
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _flow = _interopRequireDefault(require("lodash/flow"));
var _stylex2 = require("../stylex");
var _loading = require("../loading");
var _pagination = require("../pagination");
var _empty = require("../empty");
var _hooks = require("../config-provider/hooks");
var _util = require("../util");
var _useSelection = _interopRequireDefault(require("./hooks/useSelection"));
var _useSorter = _interopRequireWildcard(require("./hooks/useSorter"));
var _useFilter = _interopRequireWildcard(require("./hooks/useFilter"));
var _useTitleColumns = _interopRequireDefault(require("./hooks/useTitleColumns"));
var _usePagination = _interopRequireDefault(require("./hooks/usePagination"));
var _useBreakpoint = require("../grid/hooks/useBreakpoint");
var _ExpandIcon = _interopRequireDefault(require("./ExpandIcon"));
var _expandButton = require("../shared/components/expand-button");
var _interface = require("./interface");
Object.keys(_interface).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interface[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interface[key];
    }
  });
});
var _constants = require("./constants");
var _utils = require("./utils");
require("./style");
/**
 * @docgen
 * @public
 */

/**
 * 这一层奇奇怪怪的 wrapper 是为了让 rc-table column 能取到一个字符串的 title
 * 然后加到 th[title] 里面，
 * 如果有疑问的话，需要配合 rc-table Cell 组件一起看 hhh
 */
const FakeColumnChildren = _ref => {
  let {
    children,
    realChildren
  } = _ref;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, realChildren);
};
const Table = function (props) {
  const {
    prefixCls,
    dropdownPrefixCls,
    className,
    style,
    bordered,
    headerType = 'default',
    size,
    loading,
    rowKey = 'key',
    rowClassName,
    rowSelection,
    columns,
    dataSource,
    expandable,
    pagination,
    locale = {},
    getPopupContainer,
    sortDirections,
    showSorterTooltip = true,
    emptyText,
    renderEmpty,
    stretch,
    onChange,
    rowHoverable,
    groupConfig
  } = props;
  let {
    scroll
  } = props;
  const rcTableProps = (0, _util.omit)(props, ['className', 'style', 'columns', 'stretch']);
  if (stretch) {
    // FIXME: x 是因为目前 fixed column 的时候 fixed layout 下 empty 会滚动，max-content 会把 layout 改为 auto，之后修好可以去掉
    // y 是因为 rc-table 里面对于传了 scroll y 的情况下会判断为 fixed header，传入 css 默认值 none，没有额外含义
    scroll = {
      x: 'max-content',
      y: 'none',
      ...scroll
    };
    rcTableProps.scroll = scroll;
  }
  const mapCacheRef = _react.default.useRef({});
  const direction = (0, _hooks.useDirection)();
  const tableCssScope = (0, _hooks.useCssScope)('table');
  const prefix = prefixCls || tableCssScope();
  const dropdownCssScope = (0, _hooks.useCssScope)('menu-dropdown');
  const dropdownPrefix = dropdownPrefixCls || dropdownCssScope();
  const rawData = (0, _react.useMemo)(() => dataSource || [], [dataSource]);

  // columns 响应式
  const screens = (0, _useBreakpoint.useBreakpoint)();
  const mergedColumns = (0, _react.useMemo)(() => {
    const matched = new Set(_util.responsiveArray.filter(s => Boolean(screens[s])));
    return columns.filter(c => !c.responsive || c.responsive.some(r => matched.has(r)));
  }, [columns, screens]);
  const hasChildrenColumns = (0, _react.useMemo)(() => mergedColumns.some(item => 'children' in item), [mergedColumns]);
  let mergedData = (0, _react.useMemo)(() => dataSource || [], [dataSource]);

  // Size
  const defaultSize = (0, _hooks.useComponentSize)();
  const mergedSize = size || defaultSize;

  // Locale
  const defaultLocale = (0, _hooks.useLocale)();
  const tableLocale = {
    ...defaultLocale.components.table,
    ...locale
  };

  // Wrapper ClassNames
  const wrapperClassNames = (0, _clsx.default)(`${prefix}-wrapper`, stretch && `${prefix}-wrapper-stretch`, className);

  // Loading
  let spinProps;
  if (typeof loading === 'boolean') {
    spinProps = {
      spinning: loading
    };
  } else if (typeof loading === 'object') {
    spinProps = {
      ...loading
    };
  }

  // Expand
  const mergedExpandable = expandable || {};
  const {
    childrenColumnName = 'children'
  } = mergedExpandable;
  const expandType = _react.default.useMemo(() => {
    if (rawData.some(item => item === null || item === void 0 ? void 0 : item[childrenColumnName])) {
      return 'nest';
    }
    if (expandable && expandable.expandedRowRender) {
      return 'row';
    }
    return null;
  }, [rawData, expandable, childrenColumnName]);
  const defaultExpandIcon = (0, _ExpandIcon.default)();
  mergedExpandable.expandIcon = mergedExpandable.expandIcon || defaultExpandIcon;

  // 调整展开按钮的所在列位置
  if (expandType === 'nest' && mergedExpandable.expandIconColumnIndex === undefined) {
    mergedExpandable.expandIconColumnIndex = rowSelection ? 1 : 0;
  }

  // if (
  //   expandType === 'row' &&
  //   mergedExpandable.expandIconColumnIndex === undefined
  // ) {
  //   mergedExpandable.expandIconColumnIndex = rowSelection ? 1 : 0;
  // }

  // Indent size
  if (typeof mergedExpandable.indentSize !== 'number') {
    mergedExpandable.indentSize = 14;
  }

  // Row & Key
  const getRowKey = (0, _react.useMemo)(() => {
    if ((0, _util.isFunction)(rowKey)) {
      return rowKey;
    }
    return record => {
      var _rowKey;
      return (_rowKey = record === null || record === void 0 ? void 0 : record[rowKey]) !== null && _rowKey !== void 0 ? _rowKey : '';
    };
  }, [rowKey]);
  // get row children
  const getRowChildren = (0, _react.useCallback)(data => data === null || data === void 0 ? void 0 : data.children, []);
  const getRecordByKey = (0, _react.useMemo)(() => key => {
    if (!mapCacheRef.current || mapCacheRef.current.rawData !== rawData || mapCacheRef.current.childrenColumnName !== childrenColumnName || mapCacheRef.current.getRowKey !== getRowKey) {
      const kvMap = new Map();
      const dig = records => {
        records.forEach(record => {
          const rKey = getRowKey(record);
          kvMap.set(rKey, record);
          if (record && typeof record === 'object' && childrenColumnName in record) {
            dig(record[childrenColumnName] || []);
          }
        });
      };
      dig(rawData);
      mapCacheRef.current = {
        rawData,
        childrenColumnName,
        kvMap,
        getRowKey
      };
    }
    return mapCacheRef.current.kvMap.get(key);
  }, [rawData, childrenColumnName, getRowKey]);

  // Group
  const groupTitleRenderer = groupConfig === null || groupConfig === void 0 ? void 0 : groupConfig.groupTitleRenderer;
  const mergedGroupConfig = groupConfig || {};
  const groupRowRenderer = (0, _react.useMemo)(() => {
    if (typeof mergedGroupConfig.groupRowRenderer !== 'function') {
      return (groupKey, expanded) => {
        const title = typeof groupTitleRenderer === 'function' ? groupTitleRenderer(groupKey) : groupKey;
        return /*#__PURE__*/_react.default.createElement("div", {
          className: `${prefix}-group-title`
        }, /*#__PURE__*/_react.default.createElement(_expandButton.ExpandButton, {
          className: `${prefix}-group-title-expand-button`,
          isExpanded: expanded,
          iconType: "ExpandDownFilled",
          readonly: true
        }), title);
      };
    }
    return mergedGroupConfig.groupRowRenderer;
  }, [groupTitleRenderer, mergedGroupConfig.groupRowRenderer, prefix]);
  mergedGroupConfig.groupRowRenderer = groupRowRenderer;

  // Events
  const internalRefs = {
    body: (0, _react.useRef)()
  };
  const changeEventInfo = {};
  const triggerOnChange = function (info, action) {
    let reset = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;
    const changeInfo = {
      ...changeEventInfo,
      ...info
    };
    if (reset) {
      if ((0, _util.isFunction)(changeEventInfo.resetPagination)) {
        changeEventInfo.resetPagination();
      }

      // Reset event param
      if (changeInfo.pagination.current) {
        changeInfo.pagination.current = 1;
      }

      // Trigger pagination events
      if (pagination && pagination.onChange) {
        pagination.onChange(1, changeInfo.pagination.pageSize);
      }
    }
    if (scroll && scroll.scrollToFirstRowOnChange !== false && internalRefs.body.current) {
      internalRefs.body.current.scrollTo({
        top: 0,
        behavior: 'smooth'
      });
    }
    if (onChange) {
      onChange(changeInfo.pagination, changeInfo.filters, changeInfo.sorter, {
        currentDataSource: (0, _useFilter.getFilterData)((0, _useSorter.getSortData)(rawData, changeInfo.sorterStates, childrenColumnName), changeInfo.filterStates),
        action
      });
    }
  };

  // wrap column
  const wrapTitle = (0, _react.useCallback)((innerColumns, wrapName, otherClassName) => innerColumns.map(column => {
    let newColumn = column;
    newColumn = {
      ...newColumn,
      title: renderProps => /*#__PURE__*/_react.default.createElement("span", {
        className: (0, _clsx.default)(`${prefix}-column-${wrapName}`, otherClassName)
      }, (0, _utils.renderColumnTitle)(column.title, renderProps))
    };
    if ('children' in newColumn) {
      newColumn = {
        ...newColumn,
        children: wrapTitle(newColumn.children, wrapName, otherClassName)
      };
    }
    return newColumn;
  }), [prefix]);

  // Sorter

  const [transformSorterColumns, sortStates, sorterTitleProps, getSorters] = (0, _useSorter.default)({
    prefix,
    mergedColumns,
    onSorterChange: (sorter, sorterStates) => {
      triggerOnChange({
        sorter,
        sorterStates
      }, _interface.TableAction.SORT, false);
    },
    sortDirections: sortDirections || ['ascend', 'descend'],
    tableLocale,
    showSorterTooltip
  });
  mergedData = _react.default.useMemo(() => (0, _useSorter.getSortData)(rawData, sortStates, childrenColumnName), [rawData, sortStates, childrenColumnName]);
  changeEventInfo.sorter = getSorters();
  changeEventInfo.sorterStates = sortStates;

  // Filter

  const [transformFilterColumns, filterStates, getFilters] = (0, _useFilter.default)({
    prefix,
    dropdownPrefix,
    tableLocale,
    mergedColumns,
    onFilterChange: (filters, filterStates) => {
      triggerOnChange({
        filters,
        filterStates
      }, _interface.TableAction.FILTER, true);
    },
    getPopupContainer
  });
  mergedData = (0, _useFilter.getFilterData)(mergedData, filterStates);
  changeEventInfo.filters = getFilters();
  changeEventInfo.filterStates = filterStates;

  // Pagination

  const [mergedPagination, resetPagination, getPaginationParam] = (0, _usePagination.default)(mergedData.length, pagination, param => {
    triggerOnChange({
      pagination: param
    }, _interface.TableAction.PAGINATION, false);
  });
  changeEventInfo.pagination = pagination === false ? {} : getPaginationParam();
  changeEventInfo.resetPagination = resetPagination;

  // Render Pagination
  let topPaginationNode = null;
  let bottomPaginationNode = null;
  if (pagination !== false && mergedPagination.total) {
    // 分页器大小
    const paginationSize = mergedPagination.size ? mergedPagination.size : mergedSize === 'large' ? 'medium' : mergedSize;

    // 分页器 dom
    const renderPagination = position => /*#__PURE__*/_react.default.createElement(_pagination.Pagination, (0, _extends2.default)({
      className: `${prefix}-pagination ${prefix}-pagination-${position}`
    }, mergedPagination, {
      size: paginationSize
    }));

    // 分页器位置
    const defaultPosition = direction === 'rtl' ? 'left' : 'right';
    if (mergedPagination.position !== null && Array.isArray(mergedPagination.position)) {
      const topPos = mergedPagination.position.find(p => p.indexOf('top') !== -1);
      const bottomPos = mergedPagination.position.find(p => p.indexOf('bottom') !== -1);
      if (!topPos && !bottomPos) {
        bottomPaginationNode = renderPagination(defaultPosition);
      } else {
        if (topPos) {
          topPaginationNode = renderPagination(topPos.toLowerCase().replace('top', ''));
        }
        if (bottomPos) {
          bottomPaginationNode = renderPagination(bottomPos.toLowerCase().replace('bottom', ''));
        }
      }
    } else {
      bottomPaginationNode = renderPagination(defaultPosition);
    }
  }

  // 分页数据
  const pageData = (0, _react.useMemo)(() => {
    if (pagination === false || !mergedPagination.pageSize) {
      return mergedData;
    }
    const {
      current = 1,
      total = 0,
      pageSize = _constants.DEFAULT_PAGE_SIZE
    } = mergedPagination;
    if (mergedData.length < total) {
      if (mergedData.length > pageSize) {
        return mergedData.slice((current - 1) * pageSize, current * pageSize);
      }
      return mergedData;
    }
    return mergedData.slice((current - 1) * pageSize, current * pageSize);
  }, [mergedData, mergedPagination, pagination]);

  // Selections
  const [transformSelectionColumns, selectedKeySet] = (0, _useSelection.default)(rowSelection, {
    prefix,
    data: mergedData,
    pageData,
    getRowKey,
    getRowChildren,
    getRecordByKey,
    expandType,
    childrenColumnName,
    tableLocale,
    getPopupContainer
  });
  const internalRowClassName = (record, index, indent) => {
    let mergedRowClassName;
    if ((0, _util.isFunction)(rowClassName)) {
      mergedRowClassName = (0, _clsx.default)(rowClassName(record, index, indent));
    } else {
      mergedRowClassName = (0, _clsx.default)(rowClassName);
    }
    if (!rowHoverable || rowHoverable(record, index, indent)) {
      mergedRowClassName = (0, _clsx.default)(`${prefix}-row-hoverable`, mergedRowClassName);
    }
    return (0, _clsx.default)({
      [`${prefix}-row-selected`]: selectedKeySet.has(getRowKey(record))
    }, mergedRowClassName);
  };

  // Title
  const columnTitleProps = _react.default.useMemo(() => ({
    ...sorterTitleProps
  }), [sorterTitleProps]);
  const [transformTitleColumns] = (0, _useTitleColumns.default)(columnTitleProps);
  const renderEmptyCmp = _react.default.useMemo(() => /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(`${prefix}-empty-wrapper`)
  }, renderEmpty || /*#__PURE__*/_react.default.createElement(_empty.Empty, {
    description: emptyText || (locale === null || locale === void 0 ? void 0 : locale.emptyText)
  })), [emptyText, locale === null || locale === void 0 ? void 0 : locale.emptyText, prefix, renderEmpty]);
  const usedSize = (0, _util.getDefaultSizeAbbr)(mergedSize);
  const finalColumns = _react.default.useMemo(() => {
    /**
     * 会依次执行以下函数，封装成最终的 column title
     */

    return (0, _flow.default)([
    // 给 title 加上 一个 span 包裹
    _columns => wrapTitle(_columns, 'title', (0, _stylex.default)(_stylex2.cssTypographyHeadline)),
    // 添加 排序的 dom 结构
    transformSorterColumns,
    // 封装上筛选功能
    transformFilterColumns,
    // 整体再套一个 div 壳子
    _columns => wrapTitle(_columns, 'wrapper', ''),
    // 加上 selection 的 checkbox
    transformSelectionColumns,
    // 执行一下函数
    transformTitleColumns,
    // 封装上假的 title
    _columns => _columns.map((column, index) => {
      var _mergedColumns;
      return {
        ...column,
        title: /*#__PURE__*/_react.default.createElement(FakeColumnChildren, {
          realChildren: column.title
        }, (_mergedColumns = mergedColumns[rowSelection ? index - 1 : index]) === null || _mergedColumns === void 0 ? void 0 : _mergedColumns.title)
      };
    })])(mergedColumns);
  }, [mergedColumns, rowSelection, transformFilterColumns, transformSelectionColumns, transformSorterColumns, transformTitleColumns, wrapTitle]);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: wrapperClassNames,
    style: style
  }, /*#__PURE__*/_react.default.createElement(_loading.Loading, (0, _extends2.default)({
    spinning: false
  }, spinProps), loadingCtx => /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, topPaginationNode, /*#__PURE__*/_react.default.createElement(_rcTable.default, (0, _extends2.default)({}, rcTableProps, {
    columns: finalColumns,
    expandable: mergedExpandable,
    prefixCls: prefix,
    className: (0, _clsx.default)({
      [`${prefix}-${usedSize}`]: usedSize,
      [`${prefix}-bordered`]: bordered,
      [`${prefix}-empty`]: rawData.length === 0,
      [`${prefix}-multi-header`]: hasChildrenColumns,
      [`${prefix}-${headerType}-header`]: headerType !== 'default',
      [`${prefix}-stretch`]: stretch
    }),
    data: pageData,
    rowKey: getRowKey,
    rowClassName: internalRowClassName,
    emptyText: loadingCtx.loading ? /*#__PURE__*/_react.default.createElement("div", {
      className: `${prefix}-empty-loading`
    }) : renderEmptyCmp,
    groupConfig: mergedGroupConfig
    // Internal
    // 后期如果 rc-table 升级了，不建议使用 internal
    // 这里必须使用者 internal 参数，才能拿到内部的 scrollBodyRef，实现 scrollToFirstRowOnChange
    ,
    internalHooks: _Table.INTERNAL_HOOKS,
    internalRefs: internalRefs
  })), bottomPaginationNode)));
};
exports.Table = Table;
Table.SELECTION_ALL = _constants.SELECTION_ALL;
Table.SELECTION_INVERT = _constants.SELECTION_INVERT;
Table.Summary = _rcTable.Summary;
