// @universe-design/react/src/upload/utils/traverse-file-tree.ts
import type { RcFile } from '../interface';

interface InternalDataTransferItem extends DataTransferItem {
  isFile: boolean;
  file: (cd: (file: RcFile & { webkitRelativePath?: string }) => void) => void;
  createReader: () => any;
  fullPath: string;
  isDirectory: boolean;
  name: string;
  path: string;
}

function loopFiles(item: InternalDataTransferItem, callback: (entries: InternalDataTransferItem[]) => void) {
  const dirReader = item.createReader();
  const fileList: InternalDataTransferItem[] = [];

  function sequence() {
    dirReader?.readEntries((entries: InternalDataTransferItem[]) => {
      const entryList = Array.prototype.slice.apply(entries);
      // fileList = fileList.concat(entryList);
      fileList.push(...entryList);

      // Check if all the file has been viewed
      const isFinished = !entryList.length;

      if (isFinished) {
        callback(fileList);
      } else {
        sequence();
      }
    });
  }

  sequence();
}
const traverseFileTree = (
  files: InternalDataTransferItem[],
  callback: (files: File[]) => void,
  isAccepted: (_file: RcFile) => boolean,
) => {
  const _traverseFileTree = (item: InternalDataTransferItem, path?: string) => {
    item.path = path || '';
    if (item.isFile) {
      item.file(file => {
        if (isAccepted(file)) {
          if (item.fullPath && !file.webkitRelativePath) {
            Object.defineProperties(file, {
              webkitRelativePath: {
                writable: true,
              },
            });
            // TODO 这里有 ts 问题，webkitRelativePath 是 readonly 属性，暂时先绕过
            (
              file as Omit<File, 'webkitRelativePath'> & {
                webkitRelativePath?: string;
              }
            ).webkitRelativePath = item.fullPath.replace(/^\//, '');
            Object.defineProperties(file, {
              webkitRelativePath: {
                writable: false,
              },
            });
          }
          callback([file]);
        }
      });
    } else if (item.isDirectory) {
      loopFiles(item, (entries: InternalDataTransferItem[]) => {
        entries.forEach(entryItem => {
          _traverseFileTree(entryItem, `${path}${item.name}/`);
        });
      });
    }
  };
  files.forEach(file => {
    _traverseFileTree(
      // TODO 这里的 ts 有问题，暂时先绕过看看 js 需不需要做下处理
      file.webkitGetAsEntry() as unknown as InternalDataTransferItem,
    );
  });
};

export default traverseFileTree;
