// @byted-service/bytedtrace/lib/reporter/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AgentClient = void 0;
const tslib_1 = require("tslib");
const gConsts = tslib_1.__importStar(require("../consts"));
const status = tslib_1.__importStar(require("../status"));
const net_1 = tslib_1.__importDefault(require("net"));
const logger_1 = require("../logger");
const singleflight_1 = tslib_1.__importDefault(require("@byted-service/singleflight"));
// ByteLog protocol header:
// --------------------------------------------------------------------------------------
// |  2 bytes       |   1 byte    | 4 bytes |  4 bytes  |  1 byte       |  x bytes      |
// --------------------------------------------------------------------------------------
// |  magic number  |   version   | size    |  reserved |  log set size |  log set name |
// --------------------------------------------------------------------------------------
class AgentClient {
    constructor(statusCounter, option) {
        this.singleflight = new singleflight_1.default();
        this.options = this.fixCleintOptions(option);
        this.byteLogHeader = this.createByteLogHeader();
        this.logger = (option === null || option === void 0 ? void 0 : option.logger) || logger_1.defaultLogger;
        this.statusCounter = statusCounter;
    }
    async send(data) {
        /* istanbul ignore if */
        if (gConsts.IS_BENCHMARK) {
            return;
        }
        if (!this.socket || this.socket.destroyed) {
            await this.singleflight.do(gConsts.INIT_SOCKET_SF_KEY, async () => await this.connect());
        }
        await this.sendFull(this.getByteLogHeaderWithSize(data.length));
        await this.sendFull(data);
    }
    getByteLogHeaderWithSize(size) {
        this.byteLogHeader.writeUInt32LE(size, 3);
        return this.byteLogHeader;
    }
    async connect() {
        if (this.lastConnectTime + gConsts.RECONNECT_TIME_INTERVAL >
            Date.now()) {
            throw new Error(gConsts.CONNECT_WITHIN_TIME_INTERVAL_ERROR);
        }
        await this.singleflight.do('bytedtrace_connect_socket', async () => {
            await new Promise((resolve, reject) => {
                this.socket = new net_1.default.Socket();
                const handleError = (error) => {
                    if (this.options.netType === 'unix') {
                        this.statusCounter.recordWarningWithRateLimit(false, status.ErrUnixSocketConnectFail, 'unix socket connect failed, error: %o', error);
                    }
                    else {
                        /* istanbul ignore next */
                        this.statusCounter.recordWarningWithRateLimit(false, status.ErrTcpDialFail, 'tcp connect failed, error: %o', error);
                    }
                    this.socket.end();
                    this.socket.destroy();
                    this.lastConnectTime = Date.now();
                    reject(error);
                };
                this.socket.on('end', () => {
                    /* istanbul ignore next */
                    logger_1.debugLogger.debug('unix | tcp socket end');
                });
                this.socket.on('error', handleError);
                this.socket.on('error', async (error) => {
                    logger_1.debugLogger.debug('connect error : %o', error);
                    this.socket.end();
                    this.socket.destroy();
                });
                this.socket.connect(this.options.addr, () => {
                    this.logger.info('connect client success');
                    this.socket.removeListener('error', handleError);
                    this.lastConnectTime = Date.now();
                    resolve();
                });
                this.socket.unref();
            });
        });
    }
    async sendFull(data) {
        await new Promise((resolve, reject) => this.socket.write(data, (err) => {
            if (err) {
                /* istanbul ignore next */
                reject(err);
            }
            else {
                logger_1.debugLogger.debug('sendFull send data success');
                resolve();
            }
        }));
    }
    close() {
        if (this.socket && !this.socket.destroyed) {
            this.socket.end();
            this.socket.destroy();
        }
    }
    getByteLogSetName() {
        return process.env[gConsts.BYTELOGSETNAME] || gConsts.BYTE_LOG_SET_NAME;
    }
    createByteLogHeader() {
        const logSetName = this.getByteLogSetName();
        const logSetNameSize = Buffer.byteLength(logSetName);
        const header = Buffer.alloc(12 + logSetNameSize);
        header[0] = gConsts.BYTE_LOG_MAGIC_NUMBER;
        header[1] = gConsts.BYTE_LOG_MAGIC_NUMBER;
        header[2] = gConsts.BYTE_LOG_VERSION_BYTE;
        header[11] = logSetNameSize;
        header.write(logSetName, 12);
        return header;
    }
    fixCleintOptions(options = {}) {
        options.netType =
            typeof options.netType !== 'string'
                ? gConsts.DEFAULT_NET_TYPE
                : options.netType;
        options.addr =
            typeof options.addr !== 'object' || !options.addr
                ? gConsts.NET_TYPES[options.netType]
                : options.addr;
        return options;
    }
}
exports.AgentClient = AgentClient;
