// @byted-service/bytedtrace/lib/reporter/provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AgentProvider = void 0;
const tslib_1 = require("tslib");
/**
 * 数据上报到 agent
 */
const gloablTypes = tslib_1.__importStar(require("../types"));
const client_1 = require("./client");
const status_1 = require("../status");
class AgentProvider {
    constructor(options) {
        this.queue = [];
        this.queueLen = options.queueLen;
        this.statusCounter = options.statusCounter;
        this.sender = new client_1.AgentClient(this.statusCounter, {
            logger: options.logger,
        });
        this.status = gloablTypes.AgentProviderStatus.wait;
    }
    write(data) {
        // 超过 this.queueLen 长度或 status 状态为 close 则直接丢弃
        if (this.queue.length < this.queueLen &&
            this.status !== gloablTypes.AgentProviderStatus.close) {
            this.queue.push(data);
        }
        if ((this.status = gloablTypes.AgentProviderStatus.wait)) {
            this.run();
        }
    }
    async run() {
        while (true) {
            if (this.status === gloablTypes.AgentProviderStatus.close) {
                while (this.queue.length > 0) {
                    this.flush(this.queue.shift(), true);
                }
                this.sender.close();
                return;
            }
            if (this.queue.length <= 0) {
                this.status = gloablTypes.AgentProviderStatus.wait;
                return;
            }
            this.status = gloablTypes.AgentProviderStatus.running;
            await this.flush(this.queue.shift(), true);
        }
    }
    async flush(data, needRetry) {
        if (data.length > 0) {
            try {
                await this.sender.send(data);
            }
            catch (error) {
                if (needRetry) {
                    // 简单起见，这里与 Golang push 到 retryQueue 不同
                    await this.flush(data, false);
                }
                else {
                    this.statusCounter.recordWarningWithRateLimit(false, status_1.ErrAgentProviderFlushFail, 'agent provider sender flush data err,error :%o', error);
                }
            }
        }
    }
    close() {
        if (this.status === gloablTypes.AgentProviderStatus.wait) {
            this.sender.close();
        }
        this.status = gloablTypes.AgentProviderStatus.close;
    }
}
exports.AgentProvider = AgentProvider;
