// @universe-design/styles/cjs/vc/index.light.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.vctoken_meeting_btn_icon_schedule = exports.vctoken_meeting_btn_icon_new_meeting = exports.vctoken_meeting_btn_icon_minutes = exports.vctoken_meeting_btn_bg_toolbar = exports.vctoken_meeting_btn_bg_share_screen_sharing = exports.vctoken_meeting_btn_bg_share_screen_s = exports.vctoken_meeting_btn_bg_share_screen_pause = exports.vctoken_meeting_btn_bg_share_screen_m = exports.vctoken_meeting_btn_bg_schedule_s = exports.vctoken_meeting_btn_bg_schedule_m = exports.vctoken_meeting_btn_bg_on_gray = exports.vctoken_meeting_btn_bg_new_meeting_s = exports.vctoken_meeting_btn_bg_new_meeting_m = exports.vctoken_meeting_btn_bg_minutes_s = exports.vctoken_meeting_btn_bg_minutes_m = exports.vctoken_meeting_bg_video_off_auto_layout = exports.vctoken_meeting_bg_video_call = exports.vctoken_meeting_bg_upcoming_hover = exports.vctoken_meeting_bg_upcoming = exports.vctoken_meeting_bg_toolbar = exports.vctoken_meeting_bg_subtitle = exports.vctoken_meeting_bg_side_by_side_view = exports.vctoken_meeting_bg_share_screen_sharing = exports.vctoken_meeting_bg_share_screen_pause = exports.vctoken_meeting_bg_share_panel = exports.vctoken_meeting_bg_share_bar_immersive_mode = exports.vctoken_meeting_bg_push = exports.vctoken_meeting_bg_preview_cam_off = exports.vctoken_meeting_bg_pc_ipad_scroll_tips = exports.vctoken_meeting_bg_ongoing_hover = exports.vctoken_meeting_bg_ongoing = exports.vctoken_meeting_bg_mobile_share_bar_landscape = exports.vctoken_meeting_bg_mobile_share_bar = exports.vctoken_meeting_bg_host_tag = exports.vctoken_meeting_bg_host = exports.vctoken_meeting_bg_float_video = exports.vctoken_meeting_bg_float_transparent = exports.vctoken_meeting_bg_float_pressed = exports.vctoken_meeting_bg_float_mic = exports.vctoken_meeting_bg_float_hover = exports.vctoken_meeting_bg_float = exports.vctoken_meeting_bg_feed = exports.vctoken_meeting_bg_cohost = exports.vctoken_meeting_bg_chat_bubbles_grey_focus = exports.vctoken_meeting_bg_chat_bubbles_grey = exports.vctoken_meeting_bg_chat_bubbles_blue_focus = exports.vctoken_meeting_bg_cam_off = exports.vctoken_meeting_bg_base = exports.vctoken_meeting_bg_action_panel = exports.vctoken_meetin_bg_video_off = void 0;
exports.vctoken_vc_btn_text_selected = exports.vctoken_vc_btn_icon_share_screen = exports.vctoken_vc_btn_icon_selected = exports.vctoken_vc_btn_icon_schedule = exports.vctoken_vc_btn_icon_new_meeting = exports.vctoken_vc_btn_icon_minutes = exports.vctoken_vc_btn_icon_meeting = exports.vctoken_vc_btn_icon_livestream = exports.vctoken_vc_btn_icon_bg_share_screen = exports.vctoken_vc_btn_icon_bg_schedule = exports.vctoken_vc_btn_icon_bg_new_meeting = exports.vctoken_vc_btn_icon_bg_minutes = exports.vctoken_vc_btn_icon_bg_livestream = exports.vctoken_vc_btn_fill_selected = exports.vctoken_vc_btn_bg_share_screen = exports.vctoken_vc_btn_bg_schedule = exports.vctoken_vc_btn_bg_opacity = exports.vctoken_vc_btn_bg_new_meeting = exports.vctoken_vc_btn_bg_minutes = exports.vctoken_vc_btn_bg_livestream = exports.vctoken_vc_btn_bg_float_pressed = exports.vctoken_vc_btn_bg_float_hover = exports.vctoken_vc_btn_bg_float = exports.vctoken_vc_bg_video_call = exports.vctoken_vc_bg_subtitle = exports.vctoken_vc_bg_mobile_share_bar = exports.vctoken_vc_bg_float_video = exports.vctoken_vc_bg_float_2 = exports.vctoken_vc_bg_float = exports.vctoken_vc_bg_card_right = exports.vctoken_vc_bg_card_left = exports.vctoken_vc_bg_card_ended = exports.vctoken_vc_bg_cam_off = exports.vctoken_vc_bg_body = exports.vctoken_vc_annotate_bg_yellow = exports.vctoken_vc_annotate_bg_red = exports.vctoken_vc_annotate_bg_purple = exports.vctoken_vc_annotate_bg_green = exports.vctoken_vc_annotate_bg_blue = exports.vctoken_meeting_text_button = exports.vctoken_meeting_tag_text_cohost = exports.vctoken_meeting_tag_bg_cohost = exports.vctoken_meeting_tag_bg_blue = exports.vctoken_meeting_line_host_tag = exports.vctoken_meeting_line_divider = exports.vctoken_meeting_icon_minutes = exports.vctoken_meeting_icon_disable_on_gray_bg = exports.vctoken_meeting_fill_mask = exports.vctoken_meeting_fill_bar = exports.vctoken_meeting_btn_icon_share_screen = void 0;
exports.vctoken_vc_share_app_border = exports.vctoken_vc_shadow_xs = exports.vctoken_vc_shadow_sm = exports.vctoken_vc_list_icon_bg_upcoming_hover = exports.vctoken_vc_list_icon_bg_upcoming = exports.vctoken_vc_list_icon_bg_ongoing_hover = exports.vctoken_vc_list_icon_bg_ongoing = exports.vctoken_vc_fill_ring = exports.vctoken_vc_fill_hover = void 0;
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
exports.vctoken_meetin_bg_video_off = '#f1f3f4';
exports.vctoken_meeting_bg_action_panel = 'rgba(242, 243, 245, 0.98)';
exports.vctoken_meeting_bg_base = '#f4f5f6';
exports.vctoken_meeting_bg_cam_off = '#dee0e3';
exports.vctoken_meeting_bg_chat_bubbles_blue_focus = '#bccce5';
exports.vctoken_meeting_bg_chat_bubbles_grey = 'rgba(31, 35, 41, 0.06)';
exports.vctoken_meeting_bg_chat_bubbles_grey_focus = 'rgba(31, 35, 41, 0.12)';
exports.vctoken_meeting_bg_cohost = 'rgba(143, 149, 158, 0.8)';
exports.vctoken_meeting_bg_feed = '#1f2329';
exports.vctoken_meeting_bg_float = '#ffffff';
exports.vctoken_meeting_bg_float_hover = '#f5f6f7';
exports.vctoken_meeting_bg_float_mic = 'rgba(100, 106, 115, 0.8)';
exports.vctoken_meeting_bg_float_pressed = '#eff0f1';
exports.vctoken_meeting_bg_float_transparent = 'rgba(255, 255, 255, 0.85)';
exports.vctoken_meeting_bg_float_video = '#e8e9eb';
exports.vctoken_meeting_bg_host = 'rgba(81, 86, 93, 0.85)';
exports.vctoken_meeting_bg_host_tag = 'rgba(31, 35, 41, 0.8)';
exports.vctoken_meeting_bg_mobile_share_bar = '#eff0f1';
exports.vctoken_meeting_bg_mobile_share_bar_landscape = 'rgba(242, 243, 245, 0.9)';
exports.vctoken_meeting_bg_ongoing = '#d0f5ce';
exports.vctoken_meeting_bg_ongoing_hover = '#ffffff';
exports.vctoken_meeting_bg_pc_ipad_scroll_tips = '#ebecee';
exports.vctoken_meeting_bg_preview_cam_off = '#eaedf3';
exports.vctoken_meeting_bg_push = 'rgba(255, 255, 255, 0.94)';
exports.vctoken_meeting_bg_share_bar_immersive_mode = 'rgba(222, 224, 227, 0.9)';
exports.vctoken_meeting_bg_share_panel = '#f5f6f7';
exports.vctoken_meeting_bg_share_screen_pause = '#ffb679';
exports.vctoken_meeting_bg_share_screen_sharing = '#85dc8e';
exports.vctoken_meeting_bg_side_by_side_view = '#ebecee';
exports.vctoken_meeting_bg_subtitle = 'rgba(255, 255, 255, 0.8)';
exports.vctoken_meeting_bg_toolbar = 'rgba(255, 255, 255, 0.94)';
exports.vctoken_meeting_bg_upcoming = '#e0e9ff';
exports.vctoken_meeting_bg_upcoming_hover = '#ffffff';
exports.vctoken_meeting_bg_video_call = '#2b2f36';
exports.vctoken_meeting_bg_video_off_auto_layout = '#e9ebed';
exports.vctoken_meeting_btn_bg_minutes_m = '#e9eafb';
exports.vctoken_meeting_btn_bg_minutes_s = '#ffffff';
exports.vctoken_meeting_btn_bg_new_meeting_m = '#e0e9ff';
exports.vctoken_meeting_btn_bg_new_meeting_s = '#ffffff';
exports.vctoken_meeting_btn_bg_on_gray = '#646a73';
exports.vctoken_meeting_btn_bg_schedule_m = '#fee7cd';
exports.vctoken_meeting_btn_bg_schedule_s = '#ffffff';
exports.vctoken_meeting_btn_bg_share_screen_m = '#d0f5ce';
exports.vctoken_meeting_btn_bg_share_screen_pause = '#3b1a02';
exports.vctoken_meeting_btn_bg_share_screen_s = '#ffffff';
exports.vctoken_meeting_btn_bg_share_screen_sharing = '#022c07';
exports.vctoken_meeting_btn_bg_toolbar = '#f1f3f4';
exports.vctoken_meeting_btn_icon_minutes = '#5b65f5';
exports.vctoken_meeting_btn_icon_new_meeting = '#336df4';
exports.vctoken_meeting_btn_icon_schedule = '#ff811a';
exports.vctoken_meeting_btn_icon_share_screen = '#5cd168';
exports.vctoken_meeting_fill_bar = 'rgba(31, 35, 41, 0.05)';
exports.vctoken_meeting_fill_mask = 'rgba(0, 0, 0, 0.25)';
exports.vctoken_meeting_icon_disable_on_gray_bg = 'rgba(43, 47, 54, 0.3)';
exports.vctoken_meeting_icon_minutes = '#3ec3f7';
exports.vctoken_meeting_line_divider = '#f2f3f5';
exports.vctoken_meeting_line_host_tag = 'rgba(255, 255, 255, 0.15)';
exports.vctoken_meeting_tag_bg_blue = 'rgba(46, 101, 209, 0.6)';
exports.vctoken_meeting_tag_bg_cohost = 'rgba(143, 149, 158, 0.8)';
exports.vctoken_meeting_tag_text_cohost = '#ffffff';
exports.vctoken_meeting_text_button = '#5083fb';
exports.vctoken_vc_annotate_bg_blue = '#3370ff';
exports.vctoken_vc_annotate_bg_green = '#34c724';
exports.vctoken_vc_annotate_bg_purple = '#7f3bf5';
exports.vctoken_vc_annotate_bg_red = '#f54a45';
exports.vctoken_vc_annotate_bg_yellow = '#ffc60a';
exports.vctoken_vc_bg_body = '#f2f3f5';
exports.vctoken_vc_bg_cam_off = '#dee0e3';
exports.vctoken_vc_bg_card_ended = '#8f959e';
exports.vctoken_vc_bg_card_left = '#2e374d';
exports.vctoken_vc_bg_card_right = '#536080';
exports.vctoken_vc_bg_float = '#eff0f1';
exports.vctoken_vc_bg_float_2 = '#f5f6f7';
exports.vctoken_vc_bg_float_video = '#dee0e3';
exports.vctoken_vc_bg_mobile_share_bar = '#eff0f1';
exports.vctoken_vc_bg_subtitle = '#ffffff';
exports.vctoken_vc_bg_video_call = '#2b2f36';
exports.vctoken_vc_btn_bg_float = '#ffffff';
exports.vctoken_vc_btn_bg_float_hover = '#f2f3f5';
exports.vctoken_vc_btn_bg_float_pressed = '#eff0f1';
exports.vctoken_vc_btn_bg_livestream = '#c4f2ec';
exports.vctoken_vc_btn_bg_minutes = '#caeffc';
exports.vctoken_vc_btn_bg_new_meeting = '#e0e9ff';
exports.vctoken_vc_btn_bg_opacity = 'rgba(255, 255, 255, 0.9)';
exports.vctoken_vc_btn_bg_schedule = '#fee7cd';
exports.vctoken_vc_btn_bg_share_screen = '#d0f5ce';
exports.vctoken_vc_btn_fill_selected = 'rgba(51, 109, 244, 0.08)';
exports.vctoken_vc_btn_icon_bg_livestream = '#ffffff';
exports.vctoken_vc_btn_icon_bg_minutes = '#ffffff';
exports.vctoken_vc_btn_icon_bg_new_meeting = '#ffffff';
exports.vctoken_vc_btn_icon_bg_schedule = '#ffffff';
exports.vctoken_vc_btn_icon_bg_share_screen = '#ffffff';
exports.vctoken_vc_btn_icon_livestream = '#0f8575';
exports.vctoken_vc_btn_icon_meeting = '#336df4';
exports.vctoken_vc_btn_icon_minutes = '#047fb0';
exports.vctoken_vc_btn_icon_new_meeting = '#336df4';
exports.vctoken_vc_btn_icon_schedule = '#c25705';
exports.vctoken_vc_btn_icon_selected = '#336df4';
exports.vctoken_vc_btn_icon_share_screen = '#258832';
exports.vctoken_vc_btn_text_selected = '#336df4';
exports.vctoken_vc_fill_hover = 'rgba(31, 35, 41, 0.06)';
exports.vctoken_vc_fill_ring = '#336df4';
exports.vctoken_vc_list_icon_bg_ongoing = '#d0f5ce';
exports.vctoken_vc_list_icon_bg_ongoing_hover = '#ffffff';
exports.vctoken_vc_list_icon_bg_upcoming = '#e0e9ff';
exports.vctoken_vc_list_icon_bg_upcoming_hover = '#ffffff';
exports.vctoken_vc_shadow_sm = 'rgba(31, 35, 41, 0.5)';
exports.vctoken_vc_shadow_xs = 'rgba(31, 35, 41, 0.05)';
exports.vctoken_vc_share_app_border = '#eff0f1';
