// @universe-design/react/src/date-picker/panels/picker/Footer.tsx
import type { FC } from 'react';
import React from 'react';
import moment from 'moment';
import { useDatePickerContext } from '../../Context';
import { useCssScope, useLocale } from '../../../config-provider/hooks';
import { Button } from '../../../button';
import clsx from 'clsx';
import type { DatePickerProps } from '../../DatePicker';

export interface PickerFooterProps {
  value?: moment.Moment | null;
  /**
   * 当开启二次确认时，点击确认按钮的回调
   */
  onConfirm?: () => void;
  /**
   * 当开启二次确认时，点击取消按钮的回调
   */
  onCancel?: () => void;
  disabledDate: DatePickerProps['disabledDate'];
}

export const DatePickerFooter: FC<PickerFooterProps> = props => {
  const {
    global: globalLocale,
    components: { datePicker: locale },
  } = useLocale();
  const pickerFooterCssScope = useCssScope('picker__footer');
  const datePickerCtx = useDatePickerContext();
  const { renderExtraFooter, setValue, showToday, needConfirm, showNow } = datePickerCtx;
  function footerContent() {
    if (renderExtraFooter) {
      return (
        <div key="extra" className={pickerFooterCssScope('__extra')}>
          {renderExtraFooter(datePickerCtx)}
        </div>
      );
    }
    return (
      <>
        <div
          key="default"
          className={clsx([pickerFooterCssScope('__date'), needConfirm && pickerFooterCssScope('__date--left')])}
        >
          {showToday && (
            <Button
              type="text"
              color="primary"
              data-testid="date-picker-today"
              disabled={props.disabledDate?.(moment())}
              onClick={() => {
                setValue(moment(), !needConfirm);
              }}
            >
              {locale.today}
            </Button>
          )}

          {showNow && (
            <Button
              type="text"
              color="primary"
              data-testid="date-picker-now"
              onClick={() => {
                setValue(moment(), false);
              }}
            >
              {locale.currentTime}
            </Button>
          )}
        </div>
        {needConfirm && (
          <div key="confirm" className={pickerFooterCssScope('-confirm-buttons')}>
            <Button size="small" data-testid="date-picker-cancel-button" onClick={props.onCancel}>
              {globalLocale.cancelText}
            </Button>

            <Button
              size="small"
              type="primary"
              disabled={!props.value}
              data-testid="date-picker-confirm-button"
              onClick={props.onConfirm}
            >
              {globalLocale.okText}
            </Button>
          </div>
        )}
      </>
    );
  }

  if (!showToday && !showNow && !needConfirm && !renderExtraFooter) {
    return null;
  }
  return <div className={clsx(pickerFooterCssScope(), pickerFooterCssScope('-container'))}>{footerContent()}</div>;
};
