// @universe-design/react/src/date-picker/panels/picker/index.tsx
import clsx from 'clsx';
import type { Moment } from 'moment';
import moment from 'moment';
import React, { useCallback } from 'react';
import noop from 'lodash/noop';

import { useCssScope, useLocale } from '../../../config-provider/hooks';
import type { PanelRefProps } from '../../../shared/components/picker';
import TimePickerPanel from '../../../shared/components/picker/Comp/panels/time-panel';
import generateConfig from '../../../shared/components/picker/generate/moment';
import { DatePickerContextProvider, useDatePickerContext } from '../../Context';
import type { IFormat, MODE } from '../../interface';
import type { PickerFooterProps } from './Footer';
import { DatePickerFooter } from './Footer';
import { getTimeProps } from '../../utils/ui';
import { mergeDateAndTime } from '../../utils/date';
import { PickerContext } from '../../../shared/components/picker/Context';
import type { MonthYearSelectorProps } from '../selector/MonthYearSelector';
import type { DateCalendarUIProps } from '../DateCalendar/interface';
import useSyncState from '../../../hooks/useSyncState';
import { DateCalendar } from '../DateCalendar/DateCalendar';
import '../../style';

/**
 * @privateRemarks rangedValue不会对视图定位产生影响，value会
 * @public
 * @docgen
 * 日期选择面板的属性
 */
export interface PickerPanelProps
  extends Pick<PickerFooterProps, 'onCancel' | 'onConfirm'>,
    Pick<MonthYearSelectorProps, 'onYearClick' | 'onMonthClick' | 'onCollapsed' | 'hiddenCollapse'>,
    Omit<
      DateCalendarUIProps,
      'onPickerValueChange' | 'selectedDates' | 'selectedRange' | 'pickerValue' | 'cellRenderer' | 'previewDate'
    > {
  /**
   * 是否显示 footer
   * @defaultValue true
   */
  footer?: boolean;
  format?: IFormat;
  /**
   * 选中的值
   */
  value?: moment.Moment | null;
  /**
   * 当前视图日期发生变化时触发
   *
   * @privateRemarks
   * 1. 点击 Cell（Date、Month、Year）
   * 2. 点击 Prev Next 按钮（Date、Month、Year）
   * 3. Footer 中的『今天』
   */
  onViewDateChange?: DateCalendarUIProps['onPickerValueChange'];
  /**
   * 面板默认显示时间
   */
  defaultViewDate?: moment.Moment;
  /**
   * 自定义日期单元格的内容
   */
  dateRender?: (currentDate: moment.Moment, mode?: MODE, defaultNode?: React.ReactNode) => React.ReactNode;
}

export const isDateValueEqual = (a: moment.Moment | undefined, b: moment.Moment | undefined) => {
  if (a === b || a?.isSame(b)) {
    return true;
  }
  return false;
};
/**
 * @deprecated
 * 推荐使用 `DateCalendar` 或 `DateCalendarUI` 组件
 */
export function PickerPanel(props: PickerPanelProps) {
  const {
    className,
    style,
    value,
    mode = 'date',
    footer = true,
    defaultViewDate,
    format,
    disabledDate,
    // footerProps
    onCancel,
    onConfirm,
    // bodyProps
    onSelect,
    onDateMouseEnter: onDateMouseEnterProp,
    onViewDateChange: onPickerValueChange,
    dateRender,
    ...restBodyProps
  } = props;
  const prefixCls = useCssScope('picker')();
  const datePanelPrefixCls = `${prefixCls}-date-panel`;

  const datePickerContext = useDatePickerContext();
  const { setValue, showTime, mode: localMode, setMode: setLocalMode } = datePickerContext;

  const timeOperationRef = React.useRef<PanelRefProps>({});

  const timePanelFormat = typeof format === 'function' ? undefined : format;
  const timePickerProps = showTime
    ? getTimeProps({
        ...(typeof showTime === 'object' ? showTime : {}),
        format: timePanelFormat,
      })
    : {};
  const defaultTimeValue = timePickerProps.defaultValue || moment();

  // 当前面板切换到哪个时间点，并不是已选中的值或未提交的值
  const [pickerValue = moment(), setViewDateState] = useSyncState<moment.Moment | undefined>(value ?? defaultViewDate, {
    isEqual: isDateValueEqual,
  });

  const setViewDate = (date: Moment) => {
    onPickerValueChange?.(date, { mode: localMode });
    setViewDateState(date);
  };

  const setPickerValue = (date: React.SetStateAction<Moment | undefined>, autoCommit?: boolean) => {
    setViewDate(date as moment.Moment);
    setValue(date, autoCommit);
  };

  const onDateCalendarSelect = (date?: moment.Moment | null) => {
    if (!date) {
      // 不存在此情况
      return;
    }
    const newDate = mergeDateAndTime(date, showTime ? value ?? defaultTimeValue : null);
    onSelect?.(newDate);

    if (mode !== localMode) {
      setLocalMode(mode);
      return;
    }
  };

  const onSelectTime = useCallback(
    (time: moment.Moment) => {
      onSelect?.(time);
    },
    [onSelect],
  );

  const locale = useLocale();

  const onDateMouseEnter = useCallback(
    (v: moment.Moment) => {
      onDateMouseEnterProp?.(mergeDateAndTime(v, showTime ? value ?? defaultTimeValue : null));
    },
    [defaultTimeValue, onDateMouseEnterProp, showTime, value],
  );

  return (
    <div className={className} style={style}>
      <div className={clsx(`${datePanelPrefixCls}__wrapper`, showTime && `${datePanelPrefixCls}--withTime`)}>
        <DateCalendar
          {...restBodyProps}
          onPickerValueChange={setViewDate}
          onDateMouseEnter={onDateMouseEnter}
          value={value || undefined}
          mode={localMode}
          pickerValue={pickerValue}
          onSelect={d => onDateCalendarSelect?.(d)}
          disabledDate={disabledDate}
          {...(dateRender
            ? { cellRenderer: (date, { viewMode, defaultNode }) => dateRender(date, viewMode, defaultNode) }
            : {})}
        />
        {showTime && (
          <PickerContext.Provider
            value={{
              value: value ?? undefined,
              mode: 'time',
              setMode: noop,
              setValue,
              setOpened: noop,
              inDatePicker: true,
            }}
          >
            <TimePickerPanel<Moment>
              locale={locale.components.picker}
              {...timePickerProps}
              picker="time"
              mode="time"
              prefixCls={prefixCls}
              generateConfig={generateConfig}
              onPanelChange={args => onSelectTime(args)}
              operationRef={timeOperationRef}
            />
          </PickerContext.Provider>
        )}
      </div>
      {footer && (
        <DatePickerContextProvider value={{ ...datePickerContext, setValue: setPickerValue }}>
          <DatePickerFooter value={value} onCancel={onCancel} onConfirm={onConfirm} disabledDate={disabledDate} />
        </DatePickerContextProvider>
      )}
    </div>
  );
}

export default PickerPanel;
