// @universe-design/react/lib/shared/components/transition/FadeTransition.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FadeTransition = void 0;
var _reactUtilities = require("@universe-design/react-utilities");
var _timingFunctionConsts = require("./timingFunctionConsts");
var _hooks = require("../../../config-provider/hooks");
/* eslint-disable @typescript-eslint/no-unused-vars */

const FadeTransition = exports.FadeTransition = (0, _reactUtilities.createInOutTransition)({
  name: 'fade',
  styles: {
    appear: 'opacity: 0;',
    enter: 'opacity: 0;',
    exitDone: 'opacity: 0;',
    appearActive: 'opacity: 1;',
    enterActive: 'opacity: 1;',
    appearDone: 'opacity: 1;',
    enterDone: 'opacity: 1;',
    exit: 'opacity: 1;',
    exitActive: 'opacity: 0;'
  },
  transitions: {
    appearActive: t => `opacity ${t}ms ${_timingFunctionConsts.transitionTimingFunctionStandard}`,
    enterActive: t => `opacity ${t}ms ${_timingFunctionConsts.transitionTimingFunctionStandard}`,
    exitActive: t => `opacity ${t}ms ${_timingFunctionConsts.transitionTimingFunctionStandard}`
  },
  useCssScope: _hooks.useCssScope
});
