// @universe-design/react/lib/shared/components/transition/SlideTransition.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SlideTransition = void 0;
var _reactUtilities = require("@universe-design/react-utilities");
var _timingFunctionConsts = require("./timingFunctionConsts");
var _hooks = require("../../../config-provider/hooks");
/* eslint-disable @typescript-eslint/no-unused-vars */

const SlideTransition = exports.SlideTransition = (0, _reactUtilities.createInOutTransition)({
  name: config => `slide-${config.direction}-${config.size}`,
  styles: config => {
    const translateName = ['up', 'down'].includes(config.direction) ? 'translateY' : 'translateX';
    const inStyle = `opacity: 0; transform: ${translateName}(${['right', 'down'].includes(config.direction) ? '-' : ''}${typeof config.size === 'string' ? config.size : `${config.size}px`});`;
    const outStyle = `opacity:1; transform: ${translateName}(0px);`;
    return {
      appear: inStyle,
      enter: inStyle,
      exitDone: inStyle,
      appearActive: outStyle,
      enterActive: outStyle,
      exit: outStyle,
      exitActive: inStyle
    };
  },
  transitions: {
    appearActive: t => `opacity ${t}ms ${_timingFunctionConsts.transitionTimingFunctionStandard}, transform ${t}ms ${_timingFunctionConsts.transitionTimingFunctionStandard}`,
    enterActive: t => `opacity ${t}ms ${_timingFunctionConsts.transitionTimingFunctionStandard}, transform ${t}ms ${_timingFunctionConsts.transitionTimingFunctionStandard}`,
    exitActive: t => `opacity ${t}ms ${_timingFunctionConsts.transitionTimingFunctionStandard}, transform ${t}ms ${_timingFunctionConsts.transitionTimingFunctionStandard}`
  },
  useCssScope: _hooks.useCssScope
});
