// @universe-design/react/lib/shared/components/transition/SlideUpDownTransition.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SlideUpDownTransition = void 0;
var _reactUtilities = require("@universe-design/react-utilities");
var _timingFunctionConsts = require("./timingFunctionConsts");
var _hooks = require("../../../config-provider/hooks");
/* eslint-disable @typescript-eslint/no-unused-vars */

/**
 * 用于 select dropdown 等下拉
 */
const SlideUpDownTransition = exports.SlideUpDownTransition = (0, _reactUtilities.createInOutTransition)({
  name: _ref => {
    let {
      placement
    } = _ref;
    return `slide-up-down-${placement}`;
  },
  styles: _ref2 => {
    let {
      placement
    } = _ref2;
    const [mainAxis, crossAxis] = placement.split('-');
    let translateY = '0';
    if (mainAxis === 'bottom' || mainAxis.match(/right|left/) && crossAxis === 'start') {
      translateY = '-8px';
    } else if (mainAxis === 'top' || mainAxis.match(/right|left/) && crossAxis === 'end') {
      translateY = '8px';
    }
    return {
      appear: `opacity: 0;transform: translateY(${translateY});`,
      appearActive: 'opacity: 1;transform: translateY(0);',
      appearDone: 'opacity: 1;transform: translateY(0);',
      enter: `opacity: 0;transform: translateY(${translateY});`,
      enterActive: 'opacity: 1;transform: translateY(0);',
      enterDone: 'opacity: 1;transform: translateY(0);',
      exit: 'opacity: 1;transform: translateY(0);pointer-events: none;',
      exitActive: `opacity: 0;transform: translateY(${translateY});`,
      exitDone: `opacity: 0;transform: translateY(${translateY});pointer-events: auto;`
    };
  },
  transitions: {
    appearActive: t => `opacity ${t}ms ${_timingFunctionConsts.transitionTimingFunctionStandard}, transform ${t}ms ${_timingFunctionConsts.transitionTimingFunctionStandard}`,
    enterActive: t => `opacity ${t}ms ${_timingFunctionConsts.transitionTimingFunctionStandard}, transform ${t}ms ${_timingFunctionConsts.transitionTimingFunctionStandard}`,
    exitActive: t => `opacity ${t}ms ${_timingFunctionConsts.transitionTimingFunctionStandard}, transform ${t}ms ${_timingFunctionConsts.transitionTimingFunctionStandard}`
  },
  useCssScope: _hooks.useCssScope
});
