// @universe-design/react/lib/shared/components/transition/useAnimatedVisible.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useTransitionGroupNoopNodeRef = exports.useAnimatedVisible = void 0;
var _react = _interopRequireDefault(require("react"));
var _util = require("../../../util");
var _reactUtilities = require("@universe-design/react-utilities");
const useTransitionGroupNoopNodeRef = () =>
// https://github.com/reactjs/react-transition-group/blob/5aa3fd2d7e3354a7e42505d55af605ff44f74e2e/src/Transition.js#L325
// 必须得给个 nodeRef
// 1. 给 nodeRef 才能不 findDOMNode
// 2. nodeRef 不能为空，不然会不等待 timeout
_react.default.useRef(process.env.NODE_ENV !== "production" && (0, _reactUtilities.isClient)() ? document.createElement('div') : {});
exports.useTransitionGroupNoopNodeRef = useTransitionGroupNoopNodeRef;
const useAnimatedVisible = _ref => {
  let {
    forceRender = false,
    destroyOnClose = false,
    visible,
    onAfterOpen,
    onAfterClose
  } = _ref;
  const transitionGroupPropsWithoutListener = _react.default.useMemo(() => ({
    in: visible,
    mountOnEnter: !forceRender,
    unmountOnExit: destroyOnClose
  }), [destroyOnClose, forceRender, visible]);
  const [animatedVisible, setAnimatedVisible] = _react.default.useState(visible);
  const [enteredVisible, setEnteredVisible] = _react.default.useState(false);
  const noopRef = useTransitionGroupNoopNodeRef();

  /**
   * 模拟 getDerivedStateFromProps，如果从 visible false -> true，第一次 render  animatedVisible 还是 false，所有 child mounted 的时候父元素还是 display none，这样一些计算就没有办法做，能让用户在 content mounted 的时候拿到正确的 display 是我们的 feature
   */
  const [prevVisible, setPrevVisible] = _react.default.useState(null);
  if (prevVisible !== visible) {
    if (visible) {
      setAnimatedVisible(true);
    }
    setPrevVisible(visible);
  }
  return {
    // 因为我们在 trigger 中会把这些值作为 context 传下去，为了 context selector 能帮我们做 render 性能优化，所以都用 memo 包裹一下
    noopRef,
    transitionGroupProps: _react.default.useMemo(() => ({
      onEntered: () => {
        setEnteredVisible(true);
        (0, _util.isFunction)(onAfterOpen) && onAfterOpen();
      },
      onExit: () => {
        setEnteredVisible(false);
      },
      onExited: () => {
        setAnimatedVisible(false);
        onAfterClose === null || onAfterClose === void 0 ? void 0 : onAfterClose();
      },
      ...transitionGroupPropsWithoutListener
    }), [onAfterClose, onAfterOpen, transitionGroupPropsWithoutListener]),
    transitionGroupPropsWithoutListener,
    styleDisplay: _react.default.useMemo(() => animatedVisible ? {} : {
      display: 'none'
    }, [animatedVisible]),
    animatedVisible,
    enteredVisible
  };
};
exports.useAnimatedVisible = useAnimatedVisible;
