// @universe-design/react/lib/shared/components/transition/ZoomTransition.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ZoomTransition = void 0;
var _reactUtilities = require("@universe-design/react-utilities");
var _timingFunctionConsts = require("./timingFunctionConsts");
var _hooks = require("../../../config-provider/hooks");
/* eslint-disable @typescript-eslint/no-unused-vars */

/**
 * 用于缩放场景的动画，比如 tooltip popover popconfirm
 */
const ZoomTransition = exports.ZoomTransition = (0, _reactUtilities.createInOutTransition)({
  name: 'zoom',
  styles: {
    appear: 'opacity: 0;transform: scale(.7);',
    enter: 'opacity: 0;transform: scale(.7);',
    exitDone: 'opacity: 0;transform: scale(.7);',
    appearActive: 'opacity: 1;transform: scale(1);',
    enterActive: 'opacity: 1;transform: scale(1);',
    appearDone: 'opacity: 1;transform: scale(1);',
    enterDone: 'opacity: 1;transform: scale(1);',
    exit: 'opacity: 1;transform: scale(1);',
    exitActive: 'opacity: 0;transform: scale(0.7);'
  },
  transitions: {
    appearActive: t => `opacity ${t}ms ${_timingFunctionConsts.transitionTimingFunctionStandard}, transform ${t}ms ${_timingFunctionConsts.transitionTimingFunctionStandard}`,
    enterActive: t => `opacity ${t}ms ${_timingFunctionConsts.transitionTimingFunctionStandard}, transform ${t}ms ${_timingFunctionConsts.transitionTimingFunctionStandard}`,
    exitActive: t => `opacity ${t}ms ${_timingFunctionConsts.transitionTimingFunctionStandard}, transform ${t}ms ${_timingFunctionConsts.transitionTimingFunctionStandard}`
  },
  useCssScope: _hooks.useCssScope
});
