// @universe-design/react/es/avatar/AvatarGroup.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.u1k6fduh{cursor:pointer;}
import u12o6wmt from "./../.stylex/u12o6wmt";
//.u1xqy1su{user-select:none;}
import u1443mj0 from "./../.stylex/u1443mj0";
import clsx from 'clsx';
import * as React from 'react';
import stylex from '@universe-design/stylex';
import { COMPONENT_VERSION_ID } from '../config-provider';
import { isFunction, toFlatArray } from '../util';
import { Avatar, sizeNormalize } from './Avatar';
import { AvatarGroupContextProvider } from './AvatarGroupContext';
import { useAvatarCssScope, useAvatarGroupCssScope } from './common';
import { getAvatarGroupMoreContent } from './utils';
import { useGroupSvgMaskCls } from './AvatarGroupMask';
import { cssFontMedium } from '../stylex';
import './style';

/**
 * Avatar.Group 的属性
 * @public
 */

export const AvatarGroup = function (_ref) {
  let {
    maxCount = 5,
    children,
    avatars,
    shape = 'circle',
    size: sizeProp,
    spacing: spacingProp,
    maxProps = {},
    className,
    style,
    maxRenderer
  } = _ref;
  const avatarCssScope = useAvatarCssScope();
  const size = sizeNormalize(sizeProp);
  const spacing = spacingProp !== null && spacingProp !== void 0 ? spacingProp : -size / 4;
  const cssScope = useAvatarGroupCssScope();
  const cls = clsx(cssScope(), className, spacing < 0 && cssScope('-stacked'));
  const hollowOutCls = useGroupSvgMaskCls(size, spacing);
  let childNodes = [];
  if (children) {
    childNodes = toFlatArray(children);
  } else if (avatars) {
    childNodes = avatars.map(avatar =>
    /*#__PURE__*/
    // eslint-disable-next-line react/jsx-key
    React.createElement(Avatar, avatar));
  }
  function getAvatarsWithMargin(nodes) {
    return nodes.map((child, index) => {
      if (/*#__PURE__*/React.isValidElement(child)) {
        var _child$key;
        return /*#__PURE__*/React.cloneElement(child, {
          key: (_child$key = child.key) !== null && _child$key !== void 0 ? _child$key : `avatar-key-${index}`,
          style: {
            ...child.props.style,
            marginLeft: index === 0 ? undefined : spacing
          }
        });
      }
      return child;
    });
  }
  if (childNodes.length > maxCount) {
    const children2Show = childNodes.slice(0, maxCount - 1);
    const children2Hide = childNodes.slice(maxCount - 1, childNodes.length);
    const {
      style: maxStyle,
      className: maxClassName,
      ...restMaxProps
    } = maxProps;
    const maxAvatarNode = /*#__PURE__*/React.createElement(Avatar, _extends({
      key: "avatar-key-max",
      style: {
        marginLeft: spacing,
        ...maxStyle
      },
      shape: shape,
      size: size,
      className: clsx(avatarCssScope('-max'), maxClassName, stylex({
        "props": {
          "Bceei9c": u12o6wmt,
          "famaaq": u1443mj0
        }
      }, cssFontMedium)),
      autoScaleContent: false
    }, restMaxProps), getAvatarGroupMoreContent({
      size,
      hideCount: children2Hide.length
    }));
    const maxNode = isFunction(maxRenderer) ? maxRenderer({
      maxAvatar: maxAvatarNode,
      omittedAvatars: children2Hide.map((node, index) => ({
        node,
        index: index + children2Show.length
      }))
    }) : maxAvatarNode;
    childNodes = [...getAvatarsWithMargin(children2Show), maxNode];
  } else {
    childNodes = getAvatarsWithMargin(childNodes);
  }
  childNodes.forEach((child, index) => {
    if (/*#__PURE__*/React.isValidElement(child) && index < childNodes.length - 1) {
      childNodes[index] = /*#__PURE__*/React.cloneElement(child, {
        key: child.key,
        className: clsx(child.props.className, hollowOutCls)
      });
    }
  });
  return /*#__PURE__*/React.createElement(AvatarGroupContextProvider, {
    value: {
      shape,
      size
    }
  }, /*#__PURE__*/React.createElement("div", {
    className: cls,
    style: style
  }, childNodes));
};
Object.assign(AvatarGroup, {
  [COMPONENT_VERSION_ID]: "3.28.3"
});
