// @universe-design/react/es/avatar/squircle.js
/* eslint-disable @typescript-eslint/no-magic-numbers */
import React from 'react';
import { useCssScope } from '../config-provider/hooks';
import { useStyleSheetOnce } from '../hooks';
import { prefixToSelector } from '../util/style';
const cache = {};

// https://en.wikipedia.org/wiki/Squircle
export const getSquirclePath = size => {
  const r = size / 2;
  const step = size / 500;
  const getYorX = xOrY => Math.E ** (Math.log(r ** 4 - xOrY ** 4) / 4);
  const f2s = f => f.toFixed(2);
  const commands = [];
  commands.push(['M', `${-r} 0`]);
  for (let x = -r, prevY = 0; x <= r; x += step) {
    const y = getYorX(x);
    if (Math.abs(y - prevY) > 0.1) {
      commands.push(['L', `${f2s(x)} ${f2s(y)}`]);
      prevY = y;
    }
  }
  for (let x = r, prevY = 0; x >= -r; x -= step) {
    const y = getYorX(x);
    if (Math.abs(y - prevY) > 0.1) {
      commands.push(['L', `${f2s(x)} ${-f2s(y)}`]);
      prevY = y;
    }
  }
  commands.push(['Z']);
  return commands.map(c => c.join(' ')).join('');
};
export const SquircleAvatar = _ref => {
  let {
    size,
    bordered
  } = _ref;
  const shape = 'squircle';
  const cssKey = `${shape}-avatar-${size}`;
  if (!cache[cssKey]) {
    const svgPath = getSquirclePath(size);
    cache[cssKey] = {
      d: svgPath,
      mask: `url("data:image/svg+xml,${encodeURIComponent(`<svg width='${size}' height='${size}' viewBox='${-size / 2} ${-size / 2} ${size} ${size}' fill='none' xmlns='http://www.w3.org/2000/svg' ><path fill="black" d="${svgPath}"></path></svg>`)}") 0 0 / 100% 100% no-repeat`
    };
  }
  const {
    mask,
    d
  } = cache[cssKey];
  const cssScope = useCssScope('avatar');
  useStyleSheetOnce(`${prefixToSelector(cssScope(`-${shape}`))}-${size} {
      mask: ${mask};
    }`);
  return bordered ? /*#__PURE__*/React.createElement("svg", {
    className: cssScope(`-${shape}-border`),
    width: size,
    height: size,
    viewBox: `${-size / 2} ${-size / 2} ${size} ${size}`
  }, /*#__PURE__*/React.createElement("path", {
    strokeWidth: 2,
    d: d
  })) : null;
};
