// @universe-design/react-ai/lib-esm/onboarding-animation/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
/* eslint-disable @typescript-eslint/no-magic-numbers */
import * as React from 'react';
import animationData from './avatar-onboarding-animation.json';
import { devWarning, warning } from '@universe-design/react-utilities';
import LottieAnimation from '../shared/components/LottieAnimation';

/**
 * AI onboarding 动画组件属性
 *
 * @beta
 */

const Segments = {
  intro: [0, 147],
  defaultToAvatar: [146, 177],
  avatarToDefault: [177, 146],
  confirmAvatar: [176, 251],
  confirmDefault: [252, 351]
};
const animationGraph = [{
  from: 'none',
  to: 'default',
  segment: Segments.intro
}, {
  from: 'default',
  to: 'avatar',
  segment: Segments.defaultToAvatar
}, {
  from: 'avatar',
  to: 'default',
  segment: Segments.avatarToDefault
}, {
  from: 'avatar',
  to: 'avatarConfirmed',
  segment: Segments.confirmAvatar
}, {
  from: 'default',
  to: 'defaultConfirmed',
  segment: Segments.confirmDefault
}];

/**
 * AI onboarding 动画组件属性
 *
 * @beta
 */

/**
 * AI onboarding 动画组件
 *
 * @beta
 */
export const AIOnboardingAnimation_unstable = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    animationRef,
    ...htmlProps
  } = _ref;
  const lottieAnimationRef = React.useRef(null);
  const currentStage = React.useRef('none');
  function gotoStage(target) {
    const availablePath = animationGraph.find(edge => edge.from === currentStage.current && edge.to === target);
    if (availablePath) {
      var _lottieAnimationRef$c;
      (_lottieAnimationRef$c = lottieAnimationRef.current) === null || _lottieAnimationRef$c === void 0 ? void 0 : _lottieAnimationRef$c.playSegments([...availablePath.segment], currentStage.current === 'none');
      currentStage.current = target;
    }
    process.env.NODE_ENV !== 'production' && (process.env.NODE_ENV !== "production" ? warning(!availablePath, devWarning('OnboardingAnimation', `can't convert stage '${currentStage.current}' to stage '${target}'`)) : void 0);
  }
  React.useImperativeHandle(animationRef, () => ({
    toDefault: () => {
      gotoStage('default');
    },
    toAvatar: () => {
      gotoStage('avatar');
    },
    confirm: () => {
      if (currentStage.current === 'avatar') {
        gotoStage('avatarConfirmed');
      } else {
        gotoStage('defaultConfirmed');
      }
    }
  }), []);
  return /*#__PURE__*/React.createElement(LottieAnimation, _extends({
    ref: ref,
    animationData: animationData,
    lottieAnimationRef: lottieAnimationRef
  }, htmlProps));
});
