// @universe-design/react/lib/dialog/hooks/RefDialog.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RefModal = exports.RefConfirm = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _util = require("../../util");
var _Confirm = require("../Confirm");
var _Modal = require("../Modal");
const createRefDialog = _ref => {
  let {
    Comp,
    componentName
  } = _ref;
  const C = /*#__PURE__*/_react.default.forwardRef((_ref2, ref) => {
    let {
      defaultConfig
    } = _ref2;
    const [visible, setVisible] = _react.default.useState(true);
    const [innerConfig, setInnerConfig] = _react.default.useState(defaultConfig !== null && defaultConfig !== void 0 ? defaultConfig : {});
    _react.default.useImperativeHandle(ref, () => ({
      destroy: () => {
        setVisible(false);
      },
      update: newConfig => {
        setInnerConfig((0, _util.isFunction)(newConfig) ? newConfig : originConfig => ({
          ...originConfig,
          ...newConfig
        }));
      }
    }));
    return /*#__PURE__*/_react.default.createElement(Comp, (0, _extends2.default)({}, innerConfig, {
      visible,
      onClose: function () {
        setVisible(false);
      }
    }));
  });
  if (process.env.NODE_ENV !== "production") {
    C.displayName = componentName;
  }
  return C;
};
const RefConfirm = exports.RefConfirm = /* #__PURE__ */createRefDialog({
  Comp: _Confirm.Confirm,
  componentName: 'RefConfirm'
});
const RefModal = exports.RefModal = /* #__PURE__ */createRefDialog({
  Comp: _Modal.Modal,
  componentName: 'RefModal'
});
