// @universe-design/react/lib/dialog/hooks/useDialog.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useDialog = useDialog;
var _react = _interopRequireDefault(require("react"));
var _util = require("../../util");
var _Confirm = require("../Confirm");
var _RefDialog = require("./RefDialog");
/* eslint-disable max-nested-callbacks */
/**
 * @file useDialog
 */

/**
 * 命令式操作 react elements 的渲染
 */
const ReactElements = /*#__PURE__*/_react.default.forwardRef((_, ref) => {
  const [elements, setElements] = _react.default.useState([]);
  const addNode = _react.default.useCallback(element => {
    setElements(originElements => {
      if (originElements.includes(element)) {
        return originElements;
      }
      return [...originElements, element];
    });
    return () => {
      setElements(originElements => {
        if (originElements.includes(element)) {
          return originElements.filter(ele => ele !== element);
        }
        return originElements;
      });
    };
  }, []);
  _react.default.useImperativeHandle(ref, () => ({
    addNode
  }), [addNode]);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, elements);
});
const EMPTY_ACTION_ARRAY = [];
// 自增 id
let uid = 0;
/**
 * useDialog 可以生成能获取到 context 的 dialog 实例
 */
function useDialog() {
  const reactElementsRef = _react.default.useRef(null);
  // actionQueue 用于解决 elementsHolder 和 dialog 组件还未 render 时 open update destroy 方法不生效的问题，每次都把没执行成功的 task 放入队列，下次 render 之后延迟执行
  const [actionQueue, dispatch] = _react.default.useReducer((state, action) => {
    switch (action.type) {
      case 'enqueue':
        {
          return [...state, action.payload];
        }
      case 'dequeue':
        {
          if (!state.includes(action.payload)) {
            return state;
          }
          return state.filter(o => o !== action.payload);
        }
      default:
        {
          throw new Error('Action type is not valid.');
        }
    }
  }, EMPTY_ACTION_ARRAY);
  _react.default.useEffect(() => {
    if (!actionQueue.length) {
      return;
    }
    for (const action of actionQueue) {
      dispatch({
        type: 'dequeue',
        payload: action
      });
      action();
    }
  }, [actionQueue]);
  const openDialog = _react.default.useCallback((_ref, RefComp) => {
    let {
      onAfterClose,
      ...props
    } = _ref;
    uid += 1;
    const dialogRef = /*#__PURE__*/_react.default.createRef();
    const addNode = () => {
      var _reactElementsRef$cur;
      const unmount = (_reactElementsRef$cur = reactElementsRef.current) === null || _reactElementsRef$cur === void 0 ? void 0 : _reactElementsRef$cur.addNode(/*#__PURE__*/_react.default.createElement(RefComp, {
        key: `dialog-${uid}`,
        ref: dialogRef,
        defaultConfig: {
          ...props,
          // after close unmount 节点
          onAfterClose: () => {
            (0, _util.isFunction)(onAfterClose) && onAfterClose();
            unmount === null || unmount === void 0 ? void 0 : unmount();
          }
        }
      }));
    };
    if (reactElementsRef.current) {
      addNode();
    } else {
      dispatch({
        type: 'enqueue',
        payload: addNode
      });
    }
    const update = config => {
      if (dialogRef.current) {
        dialogRef.current.update(config);
      } else {
        dispatch({
          type: 'enqueue',
          payload: () => update(config)
        });
      }
    };
    const destroy = () => {
      if (dialogRef.current) {
        dialogRef.current.destroy();
      } else {
        dispatch({
          type: 'enqueue',
          payload: () => destroy()
        });
      }
    };
    return {
      destroy,
      update
    };
  }, []);
  const dialogInstance = _react.default.useMemo(() => ({
    confirm: props => openDialog(props, _RefDialog.RefConfirm),
    info: props => openDialog((0, _Confirm.withInfo)(props), _RefDialog.RefConfirm),
    success: props => openDialog((0, _Confirm.withSuccess)(props), _RefDialog.RefConfirm),
    error: props => openDialog((0, _Confirm.withError)(props), _RefDialog.RefConfirm),
    warn: props => openDialog((0, _Confirm.withWarn)(props), _RefDialog.RefConfirm),
    open: props => openDialog(props, _RefDialog.RefModal)
  }), [openDialog]);
  const reactElements = _react.default.useMemo(() => /*#__PURE__*/_react.default.createElement(ReactElements, {
    ref: reactElementsRef
  }), []);
  return [dialogInstance, reactElements];
}
