// @universe-design/react/es/switch/Switch.js
import _extends from "@babel/runtime/helpers/esm/extends";
import clsx from 'clsx';
import React from 'react';
import omit from 'lodash/omit';
import { useControlledState } from '@universe-design/react-utilities';
import { getSizeAbbr, devWarning, warning } from '../util';
import { useComponentSize, useCssScope } from '../config-provider/hooks';
import { SpinLoadingSvg } from '../_assets/svgs';
import { forwardHTMLRef } from '../util/forwardRef';
import './style';
export const Switch = /* #__PURE__*/Object.assign(forwardHTMLRef(function (props, ref) {
  const ctxSize = useComponentSize();
  const {
    checked: checkedProp,
    defaultChecked = false,
    loading = false,
    disabled: disabledProp = false,
    size = ctxSize === 'large' ? 'medium' : ctxSize,
    onChange,
    onClick,
    className,
    style,
    id,
    title,
    autoFocus,
    unCheckedChildren,
    checkedChildren,
    prefixCls,
    ...restProp
  } = props;
  process.env.NODE_ENV !== "production" ? warning(!unCheckedChildren && !checkedChildren, devWarning(Switch.displayName || Switch.name, 'unCheckedChildren and checkedChildren are not support.')) : void 0;
  process.env.NODE_ENV !== "production" ? warning('checked' in props || !('value' in props), devWarning('Switch', '`value` is not a valid prop, do you mean `checked`?')) : void 0;
  const cssScope = useCssScope('switch', prefixCls);
  const [checked, setChecked] = useControlledState(checkedProp !== null && checkedProp !== void 0 ? checkedProp : defaultChecked, checkedProp);
  const disabled = disabledProp || loading;
  const usedSize = getSizeAbbr(size);
  return /*#__PURE__*/React.createElement("button", _extends({}, omit(restProp, 'error'), {
    onClick: React.useCallback(e => {
      if (!disabled) {
        setChecked(_checked => !_checked, _checked => {
          onChange === null || onChange === void 0 ? void 0 : onChange(_checked, e);
          onClick === null || onClick === void 0 ? void 0 : onClick(_checked, e);
        });
      }
    }, [disabled, onChange, onClick, setChecked]),
    className: clsx(cssScope(), checked && cssScope('-checked'), disabled && cssScope('-disabled'), className, {
      [cssScope(`-${usedSize}`)]: usedSize
    }),
    style: style,
    id: id,
    type: "button",
    role: "switch",
    autoFocus: autoFocus,
    title: title,
    ref: ref,
    "aria-checked": checked
  }), /*#__PURE__*/React.createElement("div", {
    className: cssScope('__handler')
  }, loading ? /*#__PURE__*/React.createElement(SpinLoadingSvg, {
    className: cssScope('__loading')
  }) : null));
}), {
  __UD_SWITCH: true
});

/**
 * @docgen
 */

if (process.env.NODE_ENV !== "production") {
  Switch.displayName = 'Switch';
}
