// @universe-design/react/src/pagination/stories/homeless.stories.tsx
import type { Meta, StoryFn } from '@storybook/react';
import { useArgs } from '@storybook/preview-api';
import * as React from 'react';
import type { PaginationProps } from '..';
import { Pagination } from '..';
import { PerformanceLogger } from '../../_stories/common/PerformanceContent';

export default {
  title: 'Components/Pagination/临时示例',
} satisfies Meta;

export const PaginationDebug: StoryFn<PaginationProps> = props => {
  const [, update] = useArgs();
  return (
    <PerformanceLogger componentName="PaginationDebug">
      <Pagination
        {...props}
        onChange={(page, pageSize) => {
          update({
            current: page,
            pageSize,
          });
        }}
      />
    </PerformanceLogger>
  );
};

PaginationDebug.argTypes = {
  total: {
    type: 'number',
    defaultValue: 1000,
  },
  current: {
    type: 'number',
    defaultValue: 50,
  },
  pageSize: {
    type: 'number',
    defaultValue: 10,
  },
  siblingCount: {
    type: {
      name: 'object',
      value: {
        start: {
          name: 'number',
        },
        end: {
          name: 'number',
        },
      },
    },
    defaultValue: {
      start: 2,
      end: 2,
    },
  },
  boundaryCount: {
    type: {
      name: 'object',
      value: {
        start: {
          name: 'number',
        },
        end: {
          name: 'number',
        },
      },
    },
    defaultValue: {
      start: 1,
      end: 1,
    },
  },
  disabled: {
    type: 'boolean',
    defaultValue: false,
  },
  showEllipses: {
    type: 'boolean',
    defaultValue: true,
  },
};

export const SiblingCountAndBoundaryCount = () => {
  const options = React.useMemo<PaginationProps[]>(
    () => [
      {},
      { showLessItems: true },
      { siblingCount: 0 },
      { siblingCount: 1 },
      { siblingCount: 3 },
      { siblingCount: { start: 1, end: 3 } },
      { siblingCount: { start: 3, end: 1 } },
      { showEllipses: false },
      { total: 0 },
      {
        boundaryCount: { start: 1, end: 0 },
        siblingCount: { start: 10, end: 0 },
        showEllipses: false,
        total: 500,
      },
    ],
    [],
  );
  return (
    <>
      {options.map((op, i) => (
        <div key={i}>
          <h4>{JSON.stringify(op)}</h4>
          <Pagination
            // defaultCurrent={50}
            // total={2000}
            total={100}
            showSizeChanger={false}
            {...op}
          />
        </div>
      ))}
    </>
  );
};
