// @universe-design/react/src/pagination/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Basic from '../demo/0.basic';
import C10All from '../demo/10.all';
import C11BoundaryCountAndSiblingCount from '../demo/11.boundary-count-and-sibling-count';
import C2Changer from '../demo/2.changer';
import C3QuickJumper from '../demo/3.quick-jumper';
import C4Mini from '../demo/4.mini';
import C6Controlled from '../demo/6.controlled';
import C7ShowTotal from '../demo/7.show-total';
if (!(C0Basic as Story).storyName) {
  (C0Basic as Story).storyName = '0. 基础用法';
  (C0Basic as Story).parameters = {
    docs: {
      description: {
        story: '简单的分页。\n\n',
      },
    },
  };
}
if (!(C10All as Story).storyName) {
  (C10All as Story).storyName = '10. 全部展示';
  (C10All as Story).parameters = {
    docs: {
      description: {
        story: '展示所有配置选项。\n\n',
      },
    },
  };
}
if (!(C11BoundaryCountAndSiblingCount as Story).storyName) {
  (C11BoundaryCountAndSiblingCount as Story).storyName = '11. 自定义「边界页码数」和「相邻页码数」调整页码范围';
  (C11BoundaryCountAndSiblingCount as Story).parameters = {
    docs: {
      description: {
        story:
          '使用 `boundaryCount` 可以自定义分页器边界处的显示页码数量，使用 `siblingCount` 可以自定义「当前页码」相邻的显示页码数。分页器各部分对应的名词解释如下。\n\n<div style="display: flex;align: center;">\n  <div style="padding: 0 12px;">\n    <div style="white-space: pre;text-align: center;border-bottom: 1px solid var(--line-border-component);">&lt;</div>\n    <div style="text-align: center;">prev</div>\n  </div>\n  <div style="padding: 0 12px;">\n    <div style="white-space: pre;text-align: center;border-bottom: 1px solid var(--line-border-component);">1</div>\n    <div style="text-align: center;">boundaryStart</div>\n  </div>\n  <div style="padding: 0 12px;">\n    <div style="white-space: pre;text-align: center;border-bottom: 1px solid var(--line-border-component);">...</div>\n    <div style="text-align: center;">startEllipsis</div>\n  </div>\n  <div style="padding: 0 12px;">\n    <div style="white-space: pre;text-align: center;border-bottom: 1px solid var(--line-border-component);">4    5</div>\n    <div style="text-align: center;">siblingStart</div>\n  </div>\n  <div style="padding: 0 12px;">\n    <div style="white-space: pre;text-align: center;border-bottom: 1px solid var(--line-border-component);color: var(--primary-fill-default);">6</div>\n    <div style="text-align: center;">current</div>\n  </div>\n  <div style="padding: 0 12px;">\n    <div style="white-space: pre;text-align: center;border-bottom: 1px solid var(--line-border-component);">7</div>\n    <div style="text-align: center;">siblingEnd</div>\n  </div>\n  <div style="padding: 0 12px;">\n    <div style="white-space: pre;text-align: center;border-bottom: 1px solid var(--line-border-component);">...</div>\n    <div style="text-align: center;">endEllipsis</div>\n  </div>\n  <div style="padding: 0 12px;">\n    <div style="white-space: pre;text-align: center;border-bottom: 1px solid var(--line-border-component);">10    11</div>\n    <div style="text-align: center;">boundaryEnd</div>\n  </div>\n  <div style="padding: 0 12px;">\n    <div style="white-space: pre;text-align: center;border-bottom: 1px solid var(--line-border-component);">&gt;</div>\n    <div style="text-align: center;">next</div>\n  </div>\n</div>\n\n',
      },
    },
  };
}
if (!(C2Changer as Story).storyName) {
  (C2Changer as Story).storyName = '2. 改变每页展示条目';
  (C2Changer as Story).parameters = {
    docs: {
      description: {
        story: '可定义每页展示条目数量。\n\n',
      },
    },
  };
}
if (!(C3QuickJumper as Story).storyName) {
  (C3QuickJumper as Story).storyName = '3. 跳转';
  (C3QuickJumper as Story).parameters = {
    docs: {
      description: {
        story: '快速跳转到某一页。\n\n',
      },
    },
  };
}
if (!(C4Mini as Story).storyName) {
  (C4Mini as Story).storyName = '4. 迷你';
  (C4Mini as Story).parameters = {
    docs: {
      description: {
        story: '迷你版本。迷你版本不支持页码切换器和快速跳转器。\n\n',
      },
    },
  };
}
if (!(C6Controlled as Story).storyName) {
  (C6Controlled as Story).storyName = '6. 受控';
  (C6Controlled as Story).parameters = {
    docs: {
      description: {
        story: '受控制的页码。\n\n',
      },
    },
  };
}
if (!(C7ShowTotal as Story).storyName) {
  (C7ShowTotal as Story).storyName = '7. 总数';
  (C7ShowTotal as Story).parameters = {
    docs: {
      description: {
        story: '通过设置 `showTotal` 展示总共有多少数据。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Pagination/官网示例',
  tags: ['autodocs'],
};

export {
  C0Basic,
  C10All,
  C11BoundaryCountAndSiblingCount,
  C2Changer,
  C3QuickJumper,
  C4Mini,
  C6Controlled,
  C7ShowTotal,
};
