// @universe-design/colors/src/algorithm/generate-isolated-color.lark.ts
/* eslint-disable @typescript-eslint/no-magic-numbers */
import { TinyColor } from '@ctrl/tinycolor';
import type { Color } from '../types';
import { DEFAULT_COLOR_LEVELS } from '../types';

// const hueStep = 2; // 色相阶梯
// const saturationStep = 0.16; // 饱和度阶梯，浅色部分
// const saturationStep2 = 0.05; // 饱和度阶梯，深色部分
// const brightnessStep1 = 0.05; // 亮度阶梯，浅色部分
const brightnessStep2 = 0.15; // 亮度阶梯，深色部分
// const lightColorCount = 5; // 浅色数量，主色上
// const darkColorCount = 4; // 深色数量，主色下

const hueStep = 0; // 色相阶梯
const saturationStep = 0.17; // 饱和度阶梯，浅色部分
const saturationStep2 = 0.05; // 饱和度阶梯，深色部分
// const brightnessStep1 = 0.04; // 亮度阶梯，浅色部分
// const brightnessStep2 = 0.14; // 亮度阶梯，深色部分
const lightColorCount = 5; // 浅色数量，主色上
const darkColorCount = 4; // 深色数量，主色下

interface HsvObject {
  h: number;
  s: number;
  v: number;
}

function getHue(hsv: HsvObject, i: number, light?: boolean): number {
  let hue: number;
  // 根据色相不同，色相转向不同
  if (Math.round(hsv.h) >= 60 && Math.round(hsv.h) <= 230) {
    hue = light ? Math.round(hsv.h) - hueStep * i : Math.round(hsv.h) + hueStep * i;
  } else {
    hue = light ? Math.round(hsv.h) + hueStep * i : Math.round(hsv.h) - hueStep * i;
  }
  if (hue < 0) {
    hue += 360;
  } else if (hue >= 360) {
    hue -= 360;
  }
  return hue;
}

function bezier3(t: number, controllPoints: { x: number; y: number }[]) {
  const calc = (...args: number[]) =>
    (1 - t) ** 3 * args[0] + 3 * t * (1 - t) ** 2 * args[1] + 3 * t ** 2 * (1 - t) * args[2] + t ** 3 * args[3];
  return {
    x: calc(...controllPoints.map(o => o.x)),
    y: calc(...controllPoints.map(o => o.y)),
  };
}

function getSaturation(hsv: HsvObject, i: number, light?: boolean): number {
  const MIN = 0.05;
  // grey color don't change saturation
  if (hsv.h === 0 && hsv.s === 0) {
    return hsv.s;
  }
  let saturation: number;
  if (light) {
    // 0 阶
    // saturation = hsv.s - saturationStep * i;
    // 1 阶
    // const step = (hsv.s - MIN) / lightColorCount;
    // saturation = hsv.s - step * i;
    // 2 阶
    // const t = (lightColorCount - i) / lightColorCount;
    // const controllPoint = (hsv.s + MIN) / 6; // 经过点
    // saturation =
    //   (1 - t) ** 2 * MIN + 2 * t * (1 - t) * controllPoint + t ** 2 * hsv.s;
    // 3 阶
    const t = (lightColorCount - i) / lightColorCount;
    saturation = bezier3(t, [
      { x: 0, y: MIN },
      { x: 1, y: (hsv.s - MIN) * 0.05 + MIN },
      { x: 4, y: (hsv.s - MIN) * 0.8 + MIN },
      { x: 5, y: hsv.s },
    ]).y;
  } else if (i === darkColorCount) {
    saturation = hsv.s + saturationStep;
  } else {
    saturation = hsv.s + saturationStep2 * i;
  }
  // 边界值修正
  if (saturation > 1) {
    saturation = 1;
  }
  // 第一格的 s 限制在 0.06-0.1 之间
  if (light && i === lightColorCount && saturation > 0.1) {
    saturation = 0.1;
  }
  if (saturation < MIN) {
    saturation = MIN;
  }
  return Number(saturation.toFixed(2));
}

// function getSaturation2(hsv: HsvObject, i: number, light?: boolean): number {
//   return
// }

function getValue(color: TinyColor, i: number, light?: boolean): number {
  const MAX = 1;
  const hsv = color.toHsv();
  // const greyScale = color.toHsl().l;
  let value: number;
  if (light) {
    // value = hsv.v + brightnessStep1 * i;
    const step = (MAX - hsv.v) / lightColorCount;
    value = hsv.v + step * i;
  } else {
    value = hsv.v - brightnessStep2 * i;
    // 3 阶
    // console.log((hsv.s + 0.1) ** 8);
    // const t = i / darkColorCount;
    // const MIN = hsv.v - 0.57;
    // value = bezier3(t, [
    //   { x: 0, y: hsv.v },
    //   { x: 1, y: ((hsv.v - MIN) * 0.1) / (hsv.s + 0.1) ** 16 + MIN }, // (hsv.s + 0.1) ** 16 为了让 s 大的下降快一些，s 小的下降慢一点
    //   { x: 3, y: ((hsv.v - MIN) * 0.1) / (hsv.s + 0.1) ** 16 + MIN },
    //   { x: 4, y: MIN },
    // ]).y;
  }
  if (value > MAX) {
    value = MAX;
  }
  // console.log(value);
  // if (value < 0.06) {
  //   value = 0.06;
  // }
  return Number(value.toFixed(2));
}

export default function (color: string): Color[] {
  const patterns: Array<Color> = [];
  const pColor = new TinyColor(color);
  for (const [i, level] of DEFAULT_COLOR_LEVELS.entries()) {
    const isLight = i <= lightColorCount;
    const hsv = pColor.toHsv();
    const colorString: string = new TinyColor({
      h: getHue(hsv, isLight ? lightColorCount - i : i - lightColorCount, isLight),
      s: getSaturation(hsv, isLight ? lightColorCount - i : i - lightColorCount, isLight),
      v: getValue(pColor, isLight ? lightColorCount - i : i - lightColorCount, isLight),
    }).toHexString();
    patterns.push({
      level,
      value: colorString,
    });
  }
  return patterns;
}
