// @byted-service/bytedtrace/lib/confProvider/remote/configers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CONFIGER_CLASS = exports.SamplerConfiger = exports.CBConfiger = exports.GlobalConfiger = void 0;
const tslib_1 = require("tslib");
const consts = tslib_1.__importStar(require("../consts"));
// globalConfiger defines the kv store of global tracer config.
// 全局配置
class GlobalConfiger {
    constructor(kvStore) {
        this.kvStore = kvStore;
    }
    async getGlobalEnableTrace() {
        try {
            const value = await this.kvStore.getBool(consts.tccKeyGlobalEnableTracer);
            return [value, true];
        }
        catch (err) {
            /* istanbul ignore next */
            return [false, false];
        }
    }
    async getGlobalEnablePostTrace() {
        try {
            const value = await this.kvStore.getBool(consts.tccKeyGlobalEnablePostTrace);
            return [value, true];
        }
        catch (err) {
            /* istanbul ignore next */
            return [false, false];
        }
    }
    async getGlobalEnableDyedHeader() {
        try {
            const value = await this.kvStore.getBool(consts.tccKeyGlobalEnableDyedHeader);
            return [value, true];
        }
        catch (err) {
            /* istanbul ignore next */
            return [false, false];
        }
    }
    async getGlobalEnableSampling() {
        try {
            const value = await this.kvStore.getBool(consts.tccKeyGlobalEnableSampling);
            return [value, true];
        }
        catch (err) {
            /* istanbul ignore next */
            return [false, false];
        }
    }
    async getGlobalEnableReportTotalSpan() {
        try {
            const value = await this.kvStore.getBool(consts.tccKeyGlobalEnableReportTotalSpan);
            return [value, true];
        }
        catch (err) {
            /* istanbul ignore next */
            return [false, false];
        }
    }
    async getGlobalSpanMetricsLevel() {
        try {
            const value = await this.kvStore.getInt32(consts.tccKeyGlobalSpanMetricsLevel);
            return [value, true];
        }
        catch (err) {
            /* istanbul ignore next */
            return [0, false];
        }
    }
    async getGlobalCustomMetricsLevel() {
        try {
            const value = await this.kvStore.getInt32(consts.tccKeyGlobalCustomMetricsLevel);
            return [value, true];
        }
        catch (err) {
            /* istanbul ignore next */
            return [0, false];
        }
    }
    async getGlobalLightModeByComponent() {
        try {
            const value = await this.kvStore.getInterface(consts.tccKeyGlobalLightModeByComponent);
            return [value, true];
        }
        catch (err) {
            /* istanbul ignore next */
            return [undefined, false];
        }
    }
    async getGlobalRootLightModeByComponent() {
        try {
            const value = await this.kvStore.getInterface(consts.tccKeyGlobalRootLightModeByComponent);
            return [value, true];
        }
        catch (err) {
            /* istanbul ignore next */
            return [undefined, false];
        }
    }
    async getGlobalDisableThriftStyleMetricsByComponent() {
        try {
            const value = await this.kvStore.getInterface(consts.tccKeyGlobalDisableThriftStyleMetricsByComponent);
            return [value, true];
        }
        catch (err) {
            /* istanbul ignore next */
            return [undefined, false];
        }
    }
    async getGlobalRootDisableThriftStyleMetricsByComponent() {
        try {
            const value = await this.kvStore.getInterface(consts.tccKeyGlobalRootDisableThriftStyleMetricsByComponent);
            return [value, true];
        }
        catch (err) {
            /* istanbul ignore next */
            return [undefined, false];
        }
    }
}
exports.GlobalConfiger = GlobalConfiger;
// cbConfiger defines the kv store of circuit breaker config.
// 熔断器配置
class CBConfiger {
    constructor(kvStore) {
        this.kvStore = kvStore;
    }
    async getCBMaxSpanCountInTxn() {
        try {
            const value = await this.kvStore.getInt64(consts.tccKeyCBMaxSpanCountInTxn);
            return [value, true];
        }
        catch (err) {
            /* istanbul ignore next */
            return [0, false];
        }
    }
    async getCBMaxTagKVCountInObj() {
        try {
            const value = await this.kvStore.getInt64(consts.tccKeyCBMaxTagKVCountInObj);
            return [value, true];
        }
        catch (err) {
            /* istanbul ignore next */
            return [0, false];
        }
    }
    async getCBMaxEventCountInSpan() {
        try {
            const value = await this.kvStore.getInt64(consts.tccKeyCBMaxEventCountInSpan);
            return [value, true];
        }
        catch (err) {
            /* istanbul ignore next */
            return [0, false];
        }
    }
    async getCBMaxCustomMetricCountInSpan() {
        try {
            const value = await this.kvStore.getInt64(consts.tccKeyCBMaxCustomMetricCountInSpan);
            return [value, true];
        }
        catch (err) {
            /* istanbul ignore next */
            return [0, false];
        }
    }
    async getCBMaxTagValueLen() {
        try {
            const value = await this.kvStore.getInt64(consts.tccKeyCBMaxTagValueLen);
            return [value, true];
        }
        catch (err) {
            /* istanbul ignore next */
            return [0, false];
        }
    }
    async getCBMaxEventDataLen() {
        try {
            const value = await this.kvStore.getInt64(consts.tccKeyCBMaxEventDataLen);
            return [value, true];
        }
        catch (err) {
            /* istanbul ignore next */
            return [0, false];
        }
    }
    async getCBMaxTxnDataLen() {
        try {
            const value = await this.kvStore.getInt64(consts.tccKeyCBMaxTxnDataLen);
            return [value, true];
        }
        catch (err) {
            /* istanbul ignore next */
            return [0, false];
        }
    }
    async getCBMaxTxnSamplingRatePerOperation() {
        try {
            const value = await this.kvStore.getInt64(consts.tccKeyCBMaxTxnSamplingRatePerOperation);
            return [value, true];
        }
        catch (err) {
            /* istanbul ignore next */
            return [0, false];
        }
    }
    async getCBMaxPostTraceSamplingRatePerOperation() {
        try {
            const value = await this.kvStore.getFloat64(consts.tccKeyCBMaxPostTraceSamplingRatePerOperation);
            return [value, true];
        }
        catch (err) {
            /* istanbul ignore next */
            return [0, false];
        }
    }
    async getCBMaxDyeingSamplingRatePerOperation() {
        try {
            const value = await this.kvStore.getFloat64(consts.tccKeyCBMaxDyeingSamplingRatePerOperation);
            return [value, true];
        }
        catch (err) {
            /* istanbul ignore next */
            return [0, false];
        }
    }
    async getCBMaxPostTraceSamplingDetailStrategies() {
        try {
            const value = await this.kvStore.getString(consts.tccKeyCBMaxPostTraceSamplingDetailStrategies);
            const detailStrategies = JSON.parse(value);
            const detailSamplingPostTraceParam = detailStrategies.map((ds) => {
                return {
                    spanType: ds.span_type,
                    spanName: ds.span_name,
                    thresholdPerMin: ds.threshold_per_min,
                };
            });
            return [detailSamplingPostTraceParam, true];
        }
        catch (err) {
            /* istanbul ignore next */
            return [undefined, false];
        }
    }
}
exports.CBConfiger = CBConfiger;
// samplerConfiger defines the kv store of sampler config.
// 采样器配置
class SamplerConfiger {
    constructor(kvStore) {
        this.kvStore = kvStore;
    }
    async getSamplerSamplingMode() {
        try {
            const value = await this.kvStore.getString(consts.tccKeySamplerSamplingMode);
            return [value, true];
        }
        catch (err) {
            /* istanbul ignore next */
            return ['', false];
        }
    }
    async getSamplerSamplingRateProbability() {
        try {
            const value = await this.kvStore.getInt64(consts.tccKeySamplerSamplingRateProbability);
            return [value, true];
        }
        catch (err) {
            /* istanbul ignore next */
            return [0, false];
        }
    }
    async getSamplerSamplingCountPerMinAdaptive() {
        try {
            const value = await this.kvStore.getInt64(consts.tccKeySamplerSamplingCountPerMinAdaptive);
            return [value, true];
        }
        catch (err) {
            /* istanbul ignore next */
            return [0, false];
        }
    }
    async getSamplerMaxOperationCount() {
        try {
            const value = await this.kvStore.getInt64(consts.tccKeySamplerMaxOperationCount);
            return [value, true];
        }
        catch (err) {
            /* istanbul ignore next */
            return [0, false];
        }
    }
    async getSamplerDetailSamplingStrategies() {
        try {
            const value = await this.kvStore.getString(consts.tccKeySamplerDetailSamplingStrategies);
            const detailStrategies = JSON.parse(value);
            return [detailStrategies, true];
        }
        catch (err) {
            /* istanbul ignore next */
            return [undefined, false];
        }
    }
    async getSamplerConditionSamplingFilters() {
        try {
            const value = await this.kvStore.getString(consts.tccKeySamplerConditionSamplingFilters);
            const conditionSamplingFilters = JSON.parse(value);
            const conditionSamplingFiltersParam = conditionSamplingFilters.map((e) => {
                return {
                    name: e.name,
                    spanType: e.span_type,
                    conditionMetas: e.condition_metas,
                    samplingType: e.sampling_type,
                    and: e.and,
                };
            });
            return [conditionSamplingFiltersParam, true];
        }
        catch (err) {
            /* istanbul ignore next */
            return [undefined, false];
        }
    }
    async getLowerBoundSamplerExpireSecond() {
        try {
            const value = await this.kvStore.getInt64(consts.tccKeyLowerBoundSamplerExpireSecond);
            return [value, true];
        }
        catch (error) {
            /* istanbul ignore next */
            return [0, false];
        }
    }
}
exports.SamplerConfiger = SamplerConfiger;
exports.CONFIGER_CLASS = {
    [consts.NAMESPACE_TRACER]: GlobalConfiger,
    [consts.NAMESPACE_CB]: CBConfiger,
    [consts.NAMESPACE_SAMPLER]: SamplerConfiger,
};
