// @byted-service/bytedtrace/lib/confProvider/remote/kvstore.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KVStore = void 0;
const client_1 = require("./client");
// KVStore is the handler to get config from tcc.
class KVStore {
    // getString gets string value.
    async getString(key) {
        const client = await client_1.TCCClient.getDefaultClient();
        return client.getTccConfigByKey(key);
    }
    // getInt32 gets int32 value.
    async getInt32(key) {
        return this.getInt64(key);
    }
    // getInt64 gets int64 value.
    async getInt64(key) {
        const v = await this.getString(key);
        return Number(v).valueOf();
    }
    // getFloat64 gets float64 value.
    async getFloat64(key) {
        const v = await this.getString(key);
        return parseFloat(v);
    }
    // getBool gets bool value.
    async getBool(key) {
        const v = await this.getString(key);
        switch (v) {
            case 'true':
                return true;
            case 'false':
                return false;
            default:
                /* istanbul ignore next */
                return false;
        }
    }
    // getInterface gets interface value.
    async getInterface(key) {
        const v = await this.getString(key);
        return JSON.parse(v);
    }
}
exports.KVStore = KVStore;
