// @universe-design/react/es/table/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React, { useMemo, useRef, useCallback } from 'react';
import clsx from 'clsx';
import RcTable, { Summary } from '@universe-design/rc-table';
import { INTERNAL_HOOKS } from '@universe-design/rc-table/lib/Table';
import stylex from '@universe-design/stylex';
import flow from 'lodash/flow';
import { cssTypographyHeadline } from '../stylex';
import { Loading } from '../loading';
import { Pagination } from '../pagination';
import { Empty } from '../empty';
import { useComponentSize, useCssScope, useLocale, useDirection } from '../config-provider/hooks';
import { isFunction, responsiveArray, getDefaultSizeAbbr, omit } from '../util';
import useSelection from './hooks/useSelection';
import useSorter, { getSortData } from './hooks/useSorter';
import useFilter, { getFilterData } from './hooks/useFilter';
import useTitleColumns from './hooks/useTitleColumns';
import usePagination from './hooks/usePagination';
import { useBreakpoint } from '../grid/hooks/useBreakpoint';
import renderExpandIcon from './ExpandIcon';
import { ExpandButton } from '../shared/components/expand-button';
import { TableAction } from './interface';
import { SELECTION_ALL, SELECTION_INVERT, DEFAULT_PAGE_SIZE } from './constants';
import { renderColumnTitle } from './utils';
import './style';

/**
 * @docgen
 * @public
 */

/**
 * 这一层奇奇怪怪的 wrapper 是为了让 rc-table column 能取到一个字符串的 title
 * 然后加到 th[title] 里面，
 * 如果有疑问的话，需要配合 rc-table Cell 组件一起看 hhh
 */
const FakeColumnChildren = _ref => {
  let {
    children,
    realChildren
  } = _ref;
  return /*#__PURE__*/React.createElement(React.Fragment, null, realChildren);
};
export const Table = function (props) {
  const {
    prefixCls,
    dropdownPrefixCls,
    className,
    style,
    bordered,
    headerType = 'default',
    size,
    loading,
    rowKey = 'key',
    rowClassName,
    rowSelection,
    columns,
    dataSource,
    expandable,
    pagination,
    locale = {},
    getPopupContainer,
    sortDirections,
    showSorterTooltip = true,
    emptyText,
    renderEmpty,
    stretch,
    onChange,
    rowHoverable,
    groupConfig
  } = props;
  let {
    scroll
  } = props;
  const rcTableProps = omit(props, ['className', 'style', 'columns', 'stretch']);
  if (stretch) {
    // FIXME: x 是因为目前 fixed column 的时候 fixed layout 下 empty 会滚动，max-content 会把 layout 改为 auto，之后修好可以去掉
    // y 是因为 rc-table 里面对于传了 scroll y 的情况下会判断为 fixed header，传入 css 默认值 none，没有额外含义
    scroll = {
      x: 'max-content',
      y: 'none',
      ...scroll
    };
    rcTableProps.scroll = scroll;
  }
  const mapCacheRef = React.useRef({});
  const direction = useDirection();
  const tableCssScope = useCssScope('table');
  const prefix = prefixCls || tableCssScope();
  const dropdownCssScope = useCssScope('menu-dropdown');
  const dropdownPrefix = dropdownPrefixCls || dropdownCssScope();
  const rawData = useMemo(() => dataSource || [], [dataSource]);

  // columns 响应式
  const screens = useBreakpoint();
  const mergedColumns = useMemo(() => {
    const matched = new Set(responsiveArray.filter(s => Boolean(screens[s])));
    return columns.filter(c => !c.responsive || c.responsive.some(r => matched.has(r)));
  }, [columns, screens]);
  const hasChildrenColumns = useMemo(() => mergedColumns.some(item => 'children' in item), [mergedColumns]);
  let mergedData = useMemo(() => dataSource || [], [dataSource]);

  // Size
  const defaultSize = useComponentSize();
  const mergedSize = size || defaultSize;

  // Locale
  const defaultLocale = useLocale();
  const tableLocale = {
    ...defaultLocale.components.table,
    ...locale
  };

  // Wrapper ClassNames
  const wrapperClassNames = clsx(`${prefix}-wrapper`, stretch && `${prefix}-wrapper-stretch`, className);

  // Loading
  let spinProps;
  if (typeof loading === 'boolean') {
    spinProps = {
      spinning: loading
    };
  } else if (typeof loading === 'object') {
    spinProps = {
      ...loading
    };
  }

  // Expand
  const mergedExpandable = expandable || {};
  const {
    childrenColumnName = 'children'
  } = mergedExpandable;
  const expandType = React.useMemo(() => {
    if (rawData.some(item => item === null || item === void 0 ? void 0 : item[childrenColumnName])) {
      return 'nest';
    }
    if (expandable && expandable.expandedRowRender) {
      return 'row';
    }
    return null;
  }, [rawData, expandable, childrenColumnName]);
  const defaultExpandIcon = renderExpandIcon();
  mergedExpandable.expandIcon = mergedExpandable.expandIcon || defaultExpandIcon;

  // 调整展开按钮的所在列位置
  if (expandType === 'nest' && mergedExpandable.expandIconColumnIndex === undefined) {
    mergedExpandable.expandIconColumnIndex = rowSelection ? 1 : 0;
  }

  // if (
  //   expandType === 'row' &&
  //   mergedExpandable.expandIconColumnIndex === undefined
  // ) {
  //   mergedExpandable.expandIconColumnIndex = rowSelection ? 1 : 0;
  // }

  // Indent size
  if (typeof mergedExpandable.indentSize !== 'number') {
    mergedExpandable.indentSize = 14;
  }

  // Row & Key
  const getRowKey = useMemo(() => {
    if (isFunction(rowKey)) {
      return rowKey;
    }
    return record => {
      var _rowKey;
      return (_rowKey = record === null || record === void 0 ? void 0 : record[rowKey]) !== null && _rowKey !== void 0 ? _rowKey : '';
    };
  }, [rowKey]);
  // get row children
  const getRowChildren = useCallback(data => data === null || data === void 0 ? void 0 : data.children, []);
  const getRecordByKey = useMemo(() => key => {
    if (!mapCacheRef.current || mapCacheRef.current.rawData !== rawData || mapCacheRef.current.childrenColumnName !== childrenColumnName || mapCacheRef.current.getRowKey !== getRowKey) {
      const kvMap = new Map();
      const dig = records => {
        records.forEach(record => {
          const rKey = getRowKey(record);
          kvMap.set(rKey, record);
          if (record && typeof record === 'object' && childrenColumnName in record) {
            dig(record[childrenColumnName] || []);
          }
        });
      };
      dig(rawData);
      mapCacheRef.current = {
        rawData,
        childrenColumnName,
        kvMap,
        getRowKey
      };
    }
    return mapCacheRef.current.kvMap.get(key);
  }, [rawData, childrenColumnName, getRowKey]);

  // Group
  const groupTitleRenderer = groupConfig === null || groupConfig === void 0 ? void 0 : groupConfig.groupTitleRenderer;
  const mergedGroupConfig = groupConfig || {};
  const groupRowRenderer = useMemo(() => {
    if (typeof mergedGroupConfig.groupRowRenderer !== 'function') {
      return (groupKey, expanded) => {
        const title = typeof groupTitleRenderer === 'function' ? groupTitleRenderer(groupKey) : groupKey;
        return /*#__PURE__*/React.createElement("div", {
          className: `${prefix}-group-title`
        }, /*#__PURE__*/React.createElement(ExpandButton, {
          className: `${prefix}-group-title-expand-button`,
          isExpanded: expanded,
          iconType: "ExpandDownFilled",
          readonly: true
        }), title);
      };
    }
    return mergedGroupConfig.groupRowRenderer;
  }, [groupTitleRenderer, mergedGroupConfig.groupRowRenderer, prefix]);
  mergedGroupConfig.groupRowRenderer = groupRowRenderer;

  // Events
  const internalRefs = {
    body: useRef()
  };
  const changeEventInfo = {};
  const triggerOnChange = function (info, action) {
    let reset = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;
    const changeInfo = {
      ...changeEventInfo,
      ...info
    };
    if (reset) {
      if (isFunction(changeEventInfo.resetPagination)) {
        changeEventInfo.resetPagination();
      }

      // Reset event param
      if (changeInfo.pagination.current) {
        changeInfo.pagination.current = 1;
      }

      // Trigger pagination events
      if (pagination && pagination.onChange) {
        pagination.onChange(1, changeInfo.pagination.pageSize);
      }
    }
    if (scroll && scroll.scrollToFirstRowOnChange !== false && internalRefs.body.current) {
      internalRefs.body.current.scrollTo({
        top: 0,
        behavior: 'smooth'
      });
    }
    if (onChange) {
      onChange(changeInfo.pagination, changeInfo.filters, changeInfo.sorter, {
        currentDataSource: getFilterData(getSortData(rawData, changeInfo.sorterStates, childrenColumnName), changeInfo.filterStates),
        action
      });
    }
  };

  // wrap column
  const wrapTitle = useCallback((innerColumns, wrapName, otherClassName) => innerColumns.map(column => {
    let newColumn = column;
    newColumn = {
      ...newColumn,
      title: renderProps => /*#__PURE__*/React.createElement("span", {
        className: clsx(`${prefix}-column-${wrapName}`, otherClassName)
      }, renderColumnTitle(column.title, renderProps))
    };
    if ('children' in newColumn) {
      newColumn = {
        ...newColumn,
        children: wrapTitle(newColumn.children, wrapName, otherClassName)
      };
    }
    return newColumn;
  }), [prefix]);

  // Sorter

  const [transformSorterColumns, sortStates, sorterTitleProps, getSorters] = useSorter({
    prefix,
    mergedColumns,
    onSorterChange: (sorter, sorterStates) => {
      triggerOnChange({
        sorter,
        sorterStates
      }, TableAction.SORT, false);
    },
    sortDirections: sortDirections || ['ascend', 'descend'],
    tableLocale,
    showSorterTooltip
  });
  mergedData = React.useMemo(() => getSortData(rawData, sortStates, childrenColumnName), [rawData, sortStates, childrenColumnName]);
  changeEventInfo.sorter = getSorters();
  changeEventInfo.sorterStates = sortStates;

  // Filter

  const [transformFilterColumns, filterStates, getFilters] = useFilter({
    prefix,
    dropdownPrefix,
    tableLocale,
    mergedColumns,
    onFilterChange: (filters, filterStates) => {
      triggerOnChange({
        filters,
        filterStates
      }, TableAction.FILTER, true);
    },
    getPopupContainer
  });
  mergedData = getFilterData(mergedData, filterStates);
  changeEventInfo.filters = getFilters();
  changeEventInfo.filterStates = filterStates;

  // Pagination

  const [mergedPagination, resetPagination, getPaginationParam] = usePagination(mergedData.length, pagination, param => {
    triggerOnChange({
      pagination: param
    }, TableAction.PAGINATION, false);
  });
  changeEventInfo.pagination = pagination === false ? {} : getPaginationParam();
  changeEventInfo.resetPagination = resetPagination;

  // Render Pagination
  let topPaginationNode = null;
  let bottomPaginationNode = null;
  if (pagination !== false && mergedPagination.total) {
    // 分页器大小
    const paginationSize = mergedPagination.size ? mergedPagination.size : mergedSize === 'large' ? 'medium' : mergedSize;

    // 分页器 dom
    const renderPagination = position => /*#__PURE__*/React.createElement(Pagination, _extends({
      className: `${prefix}-pagination ${prefix}-pagination-${position}`
    }, mergedPagination, {
      size: paginationSize
    }));

    // 分页器位置
    const defaultPosition = direction === 'rtl' ? 'left' : 'right';
    if (mergedPagination.position !== null && Array.isArray(mergedPagination.position)) {
      const topPos = mergedPagination.position.find(p => p.indexOf('top') !== -1);
      const bottomPos = mergedPagination.position.find(p => p.indexOf('bottom') !== -1);
      if (!topPos && !bottomPos) {
        bottomPaginationNode = renderPagination(defaultPosition);
      } else {
        if (topPos) {
          topPaginationNode = renderPagination(topPos.toLowerCase().replace('top', ''));
        }
        if (bottomPos) {
          bottomPaginationNode = renderPagination(bottomPos.toLowerCase().replace('bottom', ''));
        }
      }
    } else {
      bottomPaginationNode = renderPagination(defaultPosition);
    }
  }

  // 分页数据
  const pageData = useMemo(() => {
    if (pagination === false || !mergedPagination.pageSize) {
      return mergedData;
    }
    const {
      current = 1,
      total = 0,
      pageSize = DEFAULT_PAGE_SIZE
    } = mergedPagination;
    if (mergedData.length < total) {
      if (mergedData.length > pageSize) {
        return mergedData.slice((current - 1) * pageSize, current * pageSize);
      }
      return mergedData;
    }
    return mergedData.slice((current - 1) * pageSize, current * pageSize);
  }, [mergedData, mergedPagination, pagination]);

  // Selections
  const [transformSelectionColumns, selectedKeySet] = useSelection(rowSelection, {
    prefix,
    data: mergedData,
    pageData,
    getRowKey,
    getRowChildren,
    getRecordByKey,
    expandType,
    childrenColumnName,
    tableLocale,
    getPopupContainer
  });
  const internalRowClassName = (record, index, indent) => {
    let mergedRowClassName;
    if (isFunction(rowClassName)) {
      mergedRowClassName = clsx(rowClassName(record, index, indent));
    } else {
      mergedRowClassName = clsx(rowClassName);
    }
    if (!rowHoverable || rowHoverable(record, index, indent)) {
      mergedRowClassName = clsx(`${prefix}-row-hoverable`, mergedRowClassName);
    }
    return clsx({
      [`${prefix}-row-selected`]: selectedKeySet.has(getRowKey(record))
    }, mergedRowClassName);
  };

  // Title
  const columnTitleProps = React.useMemo(() => ({
    ...sorterTitleProps
  }), [sorterTitleProps]);
  const [transformTitleColumns] = useTitleColumns(columnTitleProps);
  const renderEmptyCmp = React.useMemo(() => /*#__PURE__*/React.createElement("div", {
    className: clsx(`${prefix}-empty-wrapper`)
  }, renderEmpty || /*#__PURE__*/React.createElement(Empty, {
    description: emptyText || (locale === null || locale === void 0 ? void 0 : locale.emptyText)
  })), [emptyText, locale === null || locale === void 0 ? void 0 : locale.emptyText, prefix, renderEmpty]);
  const usedSize = getDefaultSizeAbbr(mergedSize);
  const finalColumns = React.useMemo(() => {
    /**
     * 会依次执行以下函数，封装成最终的 column title
     */

    return flow([
    // 给 title 加上 一个 span 包裹
    _columns => wrapTitle(_columns, 'title', stylex(cssTypographyHeadline)),
    // 添加 排序的 dom 结构
    transformSorterColumns,
    // 封装上筛选功能
    transformFilterColumns,
    // 整体再套一个 div 壳子
    _columns => wrapTitle(_columns, 'wrapper', ''),
    // 加上 selection 的 checkbox
    transformSelectionColumns,
    // 执行一下函数
    transformTitleColumns,
    // 封装上假的 title
    _columns => _columns.map((column, index) => {
      var _mergedColumns;
      return {
        ...column,
        title: /*#__PURE__*/React.createElement(FakeColumnChildren, {
          realChildren: column.title
        }, (_mergedColumns = mergedColumns[rowSelection ? index - 1 : index]) === null || _mergedColumns === void 0 ? void 0 : _mergedColumns.title)
      };
    })])(mergedColumns);
  }, [mergedColumns, rowSelection, transformFilterColumns, transformSelectionColumns, transformSorterColumns, transformTitleColumns, wrapTitle]);
  return /*#__PURE__*/React.createElement("div", {
    className: wrapperClassNames,
    style: style
  }, /*#__PURE__*/React.createElement(Loading, _extends({
    spinning: false
  }, spinProps), loadingCtx => /*#__PURE__*/React.createElement(React.Fragment, null, topPaginationNode, /*#__PURE__*/React.createElement(RcTable, _extends({}, rcTableProps, {
    columns: finalColumns,
    expandable: mergedExpandable,
    prefixCls: prefix,
    className: clsx({
      [`${prefix}-${usedSize}`]: usedSize,
      [`${prefix}-bordered`]: bordered,
      [`${prefix}-empty`]: rawData.length === 0,
      [`${prefix}-multi-header`]: hasChildrenColumns,
      [`${prefix}-${headerType}-header`]: headerType !== 'default',
      [`${prefix}-stretch`]: stretch
    }),
    data: pageData,
    rowKey: getRowKey,
    rowClassName: internalRowClassName,
    emptyText: loadingCtx.loading ? /*#__PURE__*/React.createElement("div", {
      className: `${prefix}-empty-loading`
    }) : renderEmptyCmp,
    groupConfig: mergedGroupConfig
    // Internal
    // 后期如果 rc-table 升级了，不建议使用 internal
    // 这里必须使用者 internal 参数，才能拿到内部的 scrollBodyRef，实现 scrollToFirstRowOnChange
    ,
    internalHooks: INTERNAL_HOOKS,
    internalRefs: internalRefs
  })), bottomPaginationNode)));
};
Table.SELECTION_ALL = SELECTION_ALL;
Table.SELECTION_INVERT = SELECTION_INVERT;
Table.Summary = Summary;
export * from './interface';
