// @universe-design/react/es/table/utils.js
import { CheckedType } from '../tree/interface';
import { isArray } from '../util';
export { CheckedType };
export function getColumnKey(column, defaultKey) {
  if ('key' in column && column.key !== undefined && column.key !== null) {
    return column.key;
  }
  if (column.dataIndex) {
    return isArray(column.dataIndex) ? column.dataIndex.join('.') : column.dataIndex;
  }
  return defaultKey;
}
export function getColumnPos(index, pos) {
  return pos ? `${pos}-${index}` : `${index}`;
}
export function renderColumnTitle(title, props) {
  if (typeof title === 'function') {
    return title(props);
  }
  return title;
}
export function traverseDataNodes(data, callback, getRowKey) {
  function processNode(node, index, parent) {
    const children = node ? node.children : data;

    // Process node if is not root
    if (node) {
      const key = getRowKey(node);
      const obj = {
        node,
        index,
        key,
        parentKey: parent.node ? getRowKey(parent.node) : '',
        depth: parent.depth + 1
      };
      callback(obj);
    }

    // Process children node
    if (children) {
      children.forEach((subNode, subIndex) => {
        processNode(subNode, subIndex, {
          node,
          key: getRowKey(node),
          depth: parent ? parent.depth + 1 : -1
        });
      });
    }
  }
  processNode(null, 0);
}

// TODO 这里类型不太对，要再仔细查一下
export function convertDataToEntities(data, getRowKey) {
  const keyEntities = {};
  traverseDataNodes(data, item => {
    const {
      node,
      index,
      key,
      parentKey,
      depth
    } = item;
    const entity = {
      node,
      index,
      key,
      depth,
      value: key
    };

    // const mergedKey = getKey(key, pos);

    keyEntities[key] = entity;

    // Fill children
    entity.parent = keyEntities[parentKey];
    if (entity.parent) {
      entity.parent.children = entity.parent.children || [];
      entity.parent.children.push(entity);
    }
  }, getRowKey);
  return {
    keyEntities
  };
}
export function isNonLeaf2(node) {
  return Boolean(node.children && node.children.length !== 0);
}
export function getCheckedType(node, checkedSet, cache) {
  const key = node.value;
  let checkedType = cache.get(key);
  if (checkedType !== undefined) {
    return checkedType;
  }
  if (checkedSet[key] || isNonLeaf2(node) && node.children.every(_node => getCheckedType(_node, checkedSet, cache) === CheckedType.Checked)) {
    checkedType = CheckedType.Checked;
    // 将所有的子元素都设置为 checked
    for (let stack = [node];;) {
      const current = stack.pop();
      if (!current) {
        break;
      }
      cache.set(current.key, CheckedType.Checked);
      stack.push(...(current.children || []));
    }
  } else if (!isNonLeaf2(node) || node.children.every(_node => getCheckedType(_node, checkedSet, cache) === CheckedType.Unchecked)) {
    checkedType = CheckedType.Unchecked;
    cache.set(key, CheckedType.Unchecked);
  } else {
    checkedType = CheckedType.Indeterminate;
    cache.set(key, CheckedType.Indeterminate);
  }
  return checkedType;
}
export function checkedSet2Cache(keyEntities, checkedSet) {
  const cache = new Map();
  for (const node of Object.values(keyEntities)) {
    getCheckedType(node, checkedSet, cache);
  }
  return cache;
}
function updateUpLevel(_node, cache, getRowKey) {
  const {
    parent
  } = _node;
  const siblings = parent === null || parent === void 0 ? void 0 : parent.children;
  if (parent && siblings) {
    const parentKey = parent.value;
    const isSiblingsAllChecked = siblings.every(o => {
      const key = getRowKey(o);
      return cache.get(key) === CheckedType.Checked;
    });
    const isSiblingsAllUnchecked = siblings.every(o => {
      const key = o.value;
      return cache.get(key) === CheckedType.Unchecked;
    });
    if (isSiblingsAllChecked) {
      cache.set(parentKey, CheckedType.Checked);
    } else if (isSiblingsAllUnchecked) {
      cache.set(parentKey, CheckedType.Unchecked);
    } else {
      cache.set(parentKey, CheckedType.Indeterminate);
    }
    updateUpLevel(parent, cache, getRowKey);
  }
}
function updateDownLevel(_node, cache, type, getRowKey) {
  if (isNonLeaf2(_node)) {
    _node.children.forEach(o => {
      const key = o.value;
      cache.set(key, type);
      updateDownLevel(o, cache, type, getRowKey);
    });
  }
}
export function check(_node, cache, checkStrictly, getRowKey) {
  cache.set(_node.value, CheckedType.Checked);
  if (!checkStrictly) {
    updateUpLevel(_node, cache, getRowKey);
    updateDownLevel(_node, cache, CheckedType.Checked, getRowKey);
  }
}
export function unCheck(_node, cache, checkStrictly, getRowKey) {
  cache.set(_node.value, CheckedType.Unchecked);
  if (!checkStrictly) {
    updateUpLevel(_node, cache, getRowKey);
    updateDownLevel(_node, cache, CheckedType.Unchecked, getRowKey);
  }
}
export function cache2CheckedSet(keyEntities, cache) {
  const checkedSet = {};
  for (const node of Object.values(keyEntities)) {
    const checkedType = cache.get(node.value);
    if (checkedType === undefined) {
      throw new Error(`checkedType of ${node.value} is not found in cache, are you sure that you have called checkedSet2Cache brefore`);
    }
    checkedSet[node.value] = checkedType === CheckedType.Checked;
  }
  return checkedSet;
}
export const selectWithShift = (lastSelectedKey, curSelectedKey, allKeys, preSelectedKeySet) => {
  let startIndex = -1;
  let endIndex = -1;
  const selectedToUnselected = preSelectedKeySet.has(curSelectedKey);
  const curSelectedKeySet = new Set(preSelectedKeySet);
  const boundaryKeys = new Set([lastSelectedKey, curSelectedKey]);
  allKeys.some((recordKey, recordIndex) => {
    if (boundaryKeys.has(recordKey)) {
      if (startIndex === -1) {
        startIndex = recordIndex;
      } else {
        endIndex = recordIndex;
        return true;
      }
    }
    return false;
  });
  if (startIndex !== endIndex && endIndex !== -1) {
    // shift 批量选择多个 key:[startIndex, endIndex]
    const selectedRangedKeys = allKeys.slice(startIndex, endIndex + 1);
    if (selectedToUnselected) {
      // 取消选择范围内的 key
      selectedRangedKeys.forEach(k => {
        if (curSelectedKeySet.has(k)) {
          curSelectedKeySet.delete(k);
        }
      });
    } else {
      // 选择范围内的 key
      selectedRangedKeys.forEach(k => {
        if (!curSelectedKeySet.has(k)) {
          curSelectedKeySet.add(k);
        }
      });
    }
  } else {
    if (selectedToUnselected) {
      curSelectedKeySet.delete(curSelectedKey);
    } else {
      curSelectedKeySet.add(curSelectedKey);
    }
  }
  return curSelectedKeySet;
};

// filter
export const serializeFilterKey = key => JSON.stringify(key);
export const serializeFilterKeys = selectedKeys => selectedKeys.map(filterKey => serializeFilterKey(filterKey));
export const unSerializeFilterKeys = serializedFilterKeys => serializedFilterKeys.map(serializedKey => JSON.parse(serializedKey));
