// @byted-service/bytedtrace/lib/metricPb.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetricPb = void 0;
const tslib_1 = require("tslib");
const utils = tslib_1.__importStar(require("./utils"));
const bytedtracePb = tslib_1.__importStar(require("./gen/bytedtrace_pb"));
class MetricPb {
    constructor(options) {
        this.startTime = utils.nowInUs();
        this.name = options.name;
        this.mType = options.mType;
        this.value = options.value;
        const tags = [];
        const keys = Object.keys(options.tagKvs);
        for (let i = 0; i < keys.length; i++) {
            const key = keys[i];
            const val = options.tagKvs[key];
            tags.push([key, val]);
        }
        this.tagKvs = options.tracer.options.breaker.buildTags(tags);
    }
    tran() {
        const mpb = new bytedtracePb.MetricProto();
        mpb.setStartTime(this.startTime);
        mpb.setMetricName(this.name);
        mpb.setMetricType(this.mType);
        mpb.setValue(this.value);
        mpb.setTagsList(this.tagKvs);
        return mpb;
    }
}
exports.MetricPb = MetricPb;
