// @byted-service/bytedtrace/lib/propagator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TextMapPropagator = void 0;
const tslib_1 = require("tslib");
/**
 * 一定要考虑兼容 trace 1.0，这里涉及到跨进程间的兼容，
 * 不能确保上下游也都是 bytedtrace
 */
const utils = tslib_1.__importStar(require("./utils"));
const span_1 = require("./span");
const tracer_1 = require("./tracer");
const logger_1 = require("./logger");
const consts_1 = require("./consts");
const jeagerLink = ':';
const w3cLink = '-';
const JaegerDebugHeader = consts_1.JAEGER_DEBUG_HEADER.toUpperCase();
const TraceContextHeaderNameJeager = consts_1.TRACE_CONTEXT_HEADER_NAME_JAEGER.toUpperCase();
const TraceBaggageHeaderName = consts_1.TRACE_BAGGAGE_HEADER_NAME.toUpperCase();
const JaegerBaggageHeader = consts_1.JAEGER_BAGGAGE_HEADER.toUpperCase();
const JaegerBaggageHeaderPrefix = consts_1.JAEGER_BAGGAGE_HEADER_PREFIX.toUpperCase();
class TextMapPropagator {
    constructor(propagatorType, logger) {
        this.propagatorType = propagatorType;
        this.logger = logger;
        this.init();
    }
    init() {
        if (this.propagatorType === consts_1.PropagatorTypes.TEXT_MAP) {
            this.encodeValue = (val) => val;
            this.decodeValue = (val) => val;
        }
        else {
            this.encodeValue = (val) => {
                try {
                    val = encodeURIComponent(val);
                }
                catch (error) /* istanbul ignore next */ {
                    this.logger.warn('encodeURIComponent error:%o', error);
                }
                return val;
            };
            this.decodeValue = (val) => {
                try {
                    val = decodeURIComponent(val);
                }
                catch (error) /* istanbul ignore next */ {
                    this.logger.warn('decodeURIComponent error:%o', error);
                }
                return val;
            };
        }
    }
    /**
     * 将 ctx 中的信息取出写入 carrier
     * 一定要考虑兼容 trace 1.0
     * @param ctx
     * @param carrier
     */
    inject(ctx, carrier) {
        if (!carrier || typeof carrier !== 'object') {
            return;
        }
        const mapDecode = this.encodeBaggage(ctx.baggage);
        carrier[consts_1.TRACE_CONTEXT_HEADER_NAME_JAEGER] = ctx.toStringJaeger();
        carrier[consts_1.JAEGER_BAGGAGE_HEADER] = mapDecode;
        carrier[consts_1.TRACER_CONTEXT_HEADER_NAME_W3C] = ctx.toStringW3C();
        carrier[consts_1.TRACE_BAGGAGE_HEADER_NAME] = mapDecode;
    }
    /**
     * 从 carrier 中获取 spanContext
     * @param carrier
     */
    extract(carrier) {
        if (!carrier || typeof carrier !== 'object') {
            return null;
        }
        if (this.containW3CSpanContext(carrier)) {
            return this.extractFromTextMapCarrier(carrier);
        }
        if (!this.containJaegerSpanContext(carrier)) {
            return null;
        }
        let ctx = null;
        const baggage = new Map();
        let debugInfo = '';
        const keys = Object.keys(carrier);
        for (let i = 0; i < keys.length; i++) {
            let key = keys[i];
            const val = carrier[key];
            key = key.toUpperCase();
            switch (key) {
                case JaegerDebugHeader: {
                    if (!debugInfo) {
                        debugInfo = this.decodeValue(val);
                    }
                    continue;
                }
                case TraceContextHeaderNameJeager: {
                    if (!ctx) {
                        ctx = this.contextFromStringJaeger(this.decodeValue(val));
                    }
                    continue;
                }
                case TraceBaggageHeaderName: {
                    this.parseCommaSeparatedMap(val, baggage, true);
                    continue;
                }
                case JaegerBaggageHeader: {
                    this.parseCommaSeparatedMap(val, baggage, false);
                    continue;
                }
            }
            if (key.startsWith(JaegerBaggageHeaderPrefix)) {
                const realKey = this.removeBaggageKeyPrefix(key);
                if (baggage.get(realKey) === undefined) {
                    baggage.set(realKey, this.decodeValue(val));
                }
            }
        }
        if (!ctx ||
            (ctx &&
                !ctx.traceID.isValid() &&
                debugInfo === '' &&
                baggage.size <= 0)) {
            /* istanbul ignore next */
            return null;
        }
        ctx.baggage = baggage;
        ctx.debugInfo = debugInfo;
        logger_1.debugLogger.debug('extract spanContext: %o', ctx);
        return ctx;
    }
    removeBaggageKeyPrefix(key) {
        return key
            .slice(consts_1.JAEGER_BAGGAGE_HEADER_PREFIX.length, key.length)
            .toLowerCase();
    }
    parseCommaSeparatedMap(value, baggage, cover) {
        if (this.propagatorType === consts_1.PropagatorTypes.HTTP_HEADER) {
            try {
                value = decodeURIComponent(value);
            }
            catch (error) /* istanbul ignore next */ {
                return baggage;
            }
        }
        const values = value.split(consts_1.COMMA);
        for (let i = 0; i < values.length; i++) {
            const v = values[i];
            const kv = v.split(consts_1.EQUAL);
            if (kv.length === 2 &&
                (cover || baggage.get(kv[0]) === undefined)) {
                baggage.set(kv[0], kv[1]);
            }
        }
        return baggage;
    }
    contextFromStringJaeger(value) {
        if (value === '' || value === undefined) {
            return null;
        }
        const parts = value.split(jeagerLink);
        if (parts.length !== 4) {
            return null;
        }
        return this.contextFromStrings(parts);
    }
    contextFromStrings(parts) {
        const ctx = new span_1.SpanContext();
        const traceID = this.traceIDFromString(parts[0]);
        if (!traceID) {
            return null;
        }
        const spanID = this.spanIDFromString(parts[1]);
        if (!spanID) {
            return null;
        }
        const parentID = this.spanIDFromString(parts[2]);
        if (!parentID) {
            return null;
        }
        ctx.traceID = traceID;
        ctx.spanID = spanID;
        ctx.parentID = parentID;
        ctx.flags = parseInt(parts[3], 10);
        return ctx;
    }
    contextFromStringW3C(value) {
        if (value === '') {
            return null;
        }
        const parts = value.split(w3cLink);
        if (parts.length !== 4) {
            return null;
        }
        return this.contextFromStringsW3c(parts);
    }
    contextFromStringsW3c(parts) {
        const ctx = new span_1.SpanContext();
        const traceID = this.traceIDFromString(parts[1]);
        if (!traceID) {
            return null;
        }
        const spanID = this.spanIDFromString(parts[2]);
        if (!spanID) {
            return null;
        }
        ctx.traceID = traceID;
        ctx.spanID = spanID;
        ctx.flags = parseInt(parts[3], 10);
        return ctx;
    }
    /**
     * 返回十进制数字字符串
     * @param s
     * @returns
     */
    spanIDFromString(s) {
        if (s.length > 16) {
            return null;
        }
        try {
            return utils.hexStrToIntStr('0x' + s);
        }
        catch (error) {
            return null;
        }
    }
    traceIDFromString(s) {
        let hi = '0';
        let lo = '0';
        const len = s.length;
        if (len > 32) {
            return null;
        }
        else if (len > 16) {
            const hiLen = len - 16;
            hi = s.slice(0, hiLen);
            lo = s.slice(hiLen, len);
        }
        else {
            lo = s;
        }
        try {
            hi = utils.hexStrToIntStr('0x' + hi);
        }
        catch (error) {
            return null;
        }
        try {
            lo = utils.hexStrToIntStr('0x' + lo);
        }
        catch (error) {
            return null;
        }
        return new tracer_1.TraceId(hi, lo);
    }
    extractFromTextMapCarrier(carrier) {
        let ctx = this.extractSpanContextFromTextMapCarrier(carrier);
        const debugInfo = this.extractDebugInfoFromTextMapCarrier(carrier);
        const baggage = new Map();
        let v = carrier[consts_1.TRACE_BAGGAGE_HEADER_NAME];
        if (v) {
            this.parseCommaSeparatedMap(v, baggage, true);
        }
        v = carrier[consts_1.JAEGER_BAGGAGE_HEADER];
        if (v) {
            this.parseCommaSeparatedMap(v, baggage, false);
        }
        if (!ctx && !debugInfo.length && !baggage.size) {
            return null;
        }
        if (!ctx) {
            ctx = new span_1.SpanContext();
        }
        ctx.debugInfo = debugInfo;
        ctx.baggage = baggage;
        return ctx;
    }
    extractDebugInfoFromTextMapCarrier(carrier) {
        const v = carrier[consts_1.TRACE_DEBUG_HEADER] || carrier[consts_1.JAEGER_DEBUG_HEADER] || '';
        if (v) {
            return this.decodeValue(v);
        }
        return '';
    }
    extractSpanContextFromTextMapCarrier(carrier) {
        let v = carrier[consts_1.TRACER_CONTEXT_HEADER_NAME_W3C];
        if (v) {
            return this.contextFromStringW3C(this.decodeValue(v));
        }
        v = carrier[consts_1.TRACE_CONTEXT_HEADER_NAME_JAEGER];
        if (v) {
            return this.contextFromStringJaeger(this.decodeValue(v));
        }
        return null;
    }
    containW3CSpanContext(carrier) {
        if (carrier[consts_1.TRACE_DEBUG_HEADER]) {
            return true;
        }
        if (carrier[consts_1.TRACER_CONTEXT_HEADER_NAME_W3C]) {
            return true;
        }
        return false;
    }
    containJaegerSpanContext(carrier) {
        if (carrier[consts_1.JAEGER_DEBUG_HEADER]) {
            return true;
        }
        if (carrier[consts_1.TRACE_CONTEXT_HEADER_NAME_JAEGER]) {
            return true;
        }
        return false;
    }
    encodeBaggage(baggage) {
        let str = '';
        for (const [key, value] of baggage) {
            if (key.includes(consts_1.EQUAL) ||
                key.includes(consts_1.COMMA) ||
                value.includes(consts_1.EQUAL) ||
                value.includes(consts_1.COMMA)) {
                return '';
            }
            const v = this.encodeValue(value || '');
            str += str ? `,${key}=${v}` : `${key}=${v}`;
        }
        return str;
    }
}
exports.TextMapPropagator = TextMapPropagator;
