// @byted-service/bytedtrace/lib/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fillExtraMetricsTags = exports.hexStrToIntStr = exports.intStrTo16HexStr = exports.transPropertyToTargetNoCheck = exports.transPropertyToTarget = exports.ifElse = exports.isSampledContext = exports.disableSampledFlag = exports.isSampledFlag = exports.nowInUs = exports.fillZeroOnHigh = exports.rangeValue = exports.contain = exports.ipv6ToBytes = exports.ipv4ToInt = exports.ipToBuffer = exports.hashToTraceID = exports.fnv32Or64 = exports.stringify = exports.sleep = void 0;
const tslib_1 = require("tslib");
const gUtils = tslib_1.__importStar(require("./consts"));
const md5_1 = tslib_1.__importDefault(require("md5"));
const tracer_1 = require("./tracer");
const PRIME_64 = 1099511628211;
const PRIME_32 = 16777619;
const FNV1_64_INIT = Number.MAX_SAFE_INTEGER;
const FNV1_32_INIT = 2166136261;
const ipv4Regex = /^(\d{1,3}\.){3,3}\d{1,3}$/;
const ipv6Regex = /^(::)?(((\d{1,3}\.){3}(\d{1,3}){1})?([0-9a-f]){0,4}:{0,2}){1,8}(::)?$/i;
const isIPv4 = (ip) => ipv4Regex.test(ip);
const isIPv6 = (ip) => ipv6Regex.test(ip);
const sleep = (ms) => {
    /* istanbul ignore next */
    return new Promise((resolve) => setTimeout(resolve, ms).unref());
};
exports.sleep = sleep;
const stringify = (value) => {
    const type = typeof value;
    switch (type) {
        case 'undefined': {
            return '';
        }
        case 'bigint': {
            return value.toString();
        }
        case 'number': {
            return value + '';
        }
        case 'boolean': {
            return value ? 'true' : 'false';
        }
        case 'object': {
            if (value instanceof Buffer) {
                return value.toString();
            }
            return JSON.stringify(value);
        }
        case 'string': {
            return value;
        }
        case 'symbol': {
            return value.toString();
        }
        case 'function': {
            return value.toString();
        }
    }
};
exports.stringify = stringify;
// hash 目的是打乱 tarceId 高位
const fnv32Or64 = (data, intType) => {
    let hash = intType === '32' ? FNV1_32_INIT : FNV1_64_INIT;
    const prime = intType === '32' ? PRIME_32 : PRIME_64;
    for (let i = 0; i < data.length; i++) {
        hash *= prime;
        hash ^= data.charCodeAt(i);
    }
    return hash > 0 ? hash : hash * -1;
};
exports.fnv32Or64 = fnv32Or64;
const hashToTraceID = (data) => {
    const h = (0, md5_1.default)(data);
    let hi = h.slice(0, 8);
    let lo = h.slice(8, 16);
    hi = (BigInt('0x' + hi) & BigInt(gUtils.MAX_INT_64)).toString(10);
    lo = (BigInt('0x' + lo) & BigInt(gUtils.MAX_INT_64)).toString(10);
    return new tracer_1.TraceId(hi, lo);
};
exports.hashToTraceID = hashToTraceID;
const ipToBuffer = (ip, buff, offset = 0) => {
    offset = ~~offset;
    let result;
    if (isIPv4(ip)) {
        result = buff || Buffer.alloc(offset + 4);
        ip.split(/\./g).map((byte) => {
            result[offset++] = parseInt(byte, 10) & 0xff;
        });
    }
    else if (isIPv6(ip)) {
        const sections = ip.split(':', 8);
        let i;
        for (i = 0; i < sections.length; i++) {
            const isv4 = isIPv4(sections[i]);
            let v4Buffer;
            if (isv4) {
                v4Buffer = (0, exports.ipToBuffer)(sections[i]);
                sections[i] = v4Buffer.slice(0, 2).toString('hex');
            }
            if (v4Buffer && ++i < 8) {
                sections.splice(i, 0, v4Buffer.slice(2, 4).toString('hex'));
            }
        }
        if (sections[0] === '') {
            while (sections.length < 8) {
                sections.unshift('0');
            }
        }
        else if (sections[sections.length - 1] === '') {
            while (sections.length < 8) {
                sections.push('0');
            }
        }
        else if (sections.length < 8) {
            const argv = [i, 1];
            for (i = 9 - sections.length; i > 0; i--) {
                argv.push('0');
            }
            sections.splice(...argv);
        }
        result = buff || Buffer.alloc(offset + 16);
        for (i = 0; i < sections.length; i++) {
            const word = parseInt(sections[i], 16);
            result[offset++] = (word >> 8) & 0xff;
            result[offset++] = word & 0xff;
        }
    }
    if (!result) {
        throw Error(`Invalid ip address: ${ip}`);
    }
    return result;
};
exports.ipToBuffer = ipToBuffer;
const ipv4ToInt = (ipStr) => {
    return (0, exports.ipToBuffer)(ipStr).readInt32LE(0);
};
exports.ipv4ToInt = ipv4ToInt;
const ipv6ToBytes = (ipv6) => {
    return (0, exports.ipToBuffer)(ipv6);
};
exports.ipv6ToBytes = ipv6ToBytes;
// 返回数组中是否包含.
const contain = (array, s) => {
    return array.indexOf(s) > -1;
};
exports.contain = contain;
const rangeValue = (min, max) => {
    if (min === max) {
        return min;
    }
    if (min > max) {
        return 0;
    }
    return parseInt(Math.random() * (max - min) + '', 10) + min;
};
exports.rangeValue = rangeValue;
/**
 * 数字长度不够 n 位，则在高位填充 0 直到长度够 n 位
 */
const fillZeroOnHigh = (n, value) => {
    let vStr = value + '';
    let gapLen = n - vStr.length;
    if (gapLen <= 0) {
        return vStr;
    }
    while (gapLen) {
        vStr = '0' + vStr;
        gapLen--;
    }
    return vStr;
};
exports.fillZeroOnHigh = fillZeroOnHigh;
const uptime = Date.now() * 1000;
const hrtime = process.hrtime();
const nowInUs = () => {
    const [sec, nsec] = process.hrtime(hrtime);
    const mms = Math.floor(sec * 1e6 + nsec / 1e3);
    return uptime + mms;
};
exports.nowInUs = nowInUs;
const isSampledFlag = (f) => {
    return (f & gUtils.SAMPLED_FLAGS) === gUtils.SAMPLED_FLAGS;
};
exports.isSampledFlag = isSampledFlag;
const disableSampledFlag = (f) => {
    return f & gUtils.DISABLE_SAMPLE;
};
exports.disableSampledFlag = disableSampledFlag;
const isSampledContext = (ctx) => {
    if (ctx.noop()) {
        return false;
    }
    if (ctx.isDebug()) {
        return true;
    }
    return (ctx.flags & gUtils.SAMPLED_FLAGS) === gUtils.SAMPLED_FLAGS;
};
exports.isSampledContext = isSampledContext;
// ifElse returns newValue if status is true else oldValue.
function ifElse(status, newValue, oldValue) {
    return status === true && newValue !== undefined ? newValue : oldValue;
}
exports.ifElse = ifElse;
/**
 * 如果 target 不存在属性 k 那么将 obj 的 k 属性值赋值给 target 的 k 属性
 * @param target
 * @param obj
 * @returns
 */
function transPropertyToTarget(target = {}, obj) {
    const keys = Object.keys(obj);
    for (let i = 0; i < keys.length; i++) {
        const k = keys[i];
        if (target[k] === undefined) {
            target[k] = obj[k];
        }
    }
    return target;
}
exports.transPropertyToTarget = transPropertyToTarget;
/**
 * 直接覆盖属性
 * @param target
 * @param obj
 * @returns
 */
function transPropertyToTargetNoCheck(target, obj) {
    return Object.assign(target, obj);
}
exports.transPropertyToTargetNoCheck = transPropertyToTargetNoCheck;
/**
 * 10 进制 int 字符串转成 16 进制数字字符串
 * @param num
 * @returns
 */
function intStrTo16HexStr(num) {
    return BigInt(num).toString(16);
}
exports.intStrTo16HexStr = intStrTo16HexStr;
/**
 * 将字符串转成 10 进制 int 字符
 * @param num
 * @returns
 */
function hexStrToIntStr(num) {
    return BigInt(num).toString(10);
}
exports.hexStrToIntStr = hexStrToIntStr;
function fillExtraMetricsTags(spanTags, registeredTags, simplify) {
    const tags = {};
    if (simplify) {
        registeredTags.forEach((k) => {
            tags[k] = '-';
        });
    }
    else {
        registeredTags.forEach((k) => {
            const tag = spanTags.find((v) => v[0] === k);
            if (tag) {
                tags[k] = (0, exports.stringify)(tag[1]);
            }
            else {
                tags[k] = '-';
            }
        });
    }
    return tags;
}
exports.fillExtraMetricsTags = fillExtraMetricsTags;
