// @byted-service/tcc-general/lib/cacheNode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const bconfig_1 = tslib_1.__importDefault(require("./bconfig"));
const node_1 = tslib_1.__importDefault(require("./node"));
const consts_1 = require("./consts");
const error_1 = require("./error");
const CACHEES = {};
const REFRESH_NODE_COUNT = 100;
const MAX_DONT_CACHE_COUNT = 3;
const CACHE_REQ = {};
class CacheNode {
    static create(cluster, options) {
        if (CACHEES[cluster]) {
            return CACHEES[cluster];
        }
        const cache = new CacheNode(cluster, options);
        cache.refresh();
        CACHEES[cluster] = cache;
        return cache;
    }
    constructor(cluster, options) {
        this.getCount = 0;
        this.getSuccCount = 0;
        this.cluster = cluster;
        this.cacheDataSet = {};
        this.logger = options.logger;
        this.bconfigClient = new bconfig_1.default(this.logger, cluster, options.getTimeout);
    }
    getRandom(n) {
        return (Math.random() + 1) * n;
    }
    async get(nodeName, cacheTime) {
        // get from cache
        const cachedNodeData = this._get(nodeName);
        if (cachedNodeData && !cachedNodeData.expired()) {
            return {
                configNodeData: cachedNodeData,
                getType: consts_1.GetType.getTypeCache,
            };
        }
        const versionCode = cachedNodeData
            ? cachedNodeData.getVersionCode()
            : '';
        // get from remote
        let data;
        let getType = consts_1.GetType.getTypeServer;
        let lastErr;
        const cacheKey = `${this.cluster}-${nodeName}`;
        try {
            // only one request will be send
            if (!CACHE_REQ[cacheKey]) {
                CACHE_REQ[cacheKey] = this.getFromRemote(nodeName, versionCode);
            }
            data = await CACHE_REQ[cacheKey];
        }
        catch (error) {
            // if data has been cached, should return cached data
            // tslint:disable-next-line: triple-equals
            if (error.code != error_1.TccErrorCode.errKeyNotFound && cachedNodeData) {
                cachedNodeData.renew();
                getType =
                    error.code === error_1.TccErrorCode.errNotModified
                        ? getType
                        : consts_1.GetType.getTypeStaleCache;
                return {
                    configNodeData: cachedNodeData,
                    getType,
                };
            }
            lastErr = error;
        }
        finally {
            delete CACHE_REQ[cacheKey];
        }
        const res = {
            configNodeData: new node_1.default({
                data,
                cacheTime,
                nodeName,
                lastErr,
            }),
            getType,
        };
        // if err != ErrNotFound, and exceed MaxDontCacheCount or has get successfully ever, cache err
        if (this.getCount > MAX_DONT_CACHE_COUNT || this.getSuccCount > 0) {
            this.set(nodeName, res.configNodeData);
        }
        return res;
    }
    set(nodeName, nodeData) {
        this.cacheDataSet[nodeName] = nodeData;
    }
    _get(nodeName) {
        return this.cacheDataSet[nodeName];
    }
    getAll() {
        return Object.assign({}, this.cacheDataSet);
    }
    async getFromRemote(nodeName, versionCode) {
        this.getCount++;
        try {
            const respData = await this.bconfigClient.getRemote(nodeName, versionCode);
            this.getSuccCount++;
            return respData;
        }
        catch (error) {
            if (
            // tslint:disable-next-line: triple-equals
            error.code == error_1.TccErrorCode.errKeyNotFound ||
                error.code === error_1.TccErrorCode.errNotModified) {
                this.getSuccCount++;
            }
            throw error;
        }
    }
    async refresh() {
        // refresh time in 5-10s
        let nextRefreshInterval = this.getRandom(5) * 1000;
        while (true) {
            await (0, consts_1.sleep)(nextRefreshInterval);
            nextRefreshInterval = this.getRandom(5) * 1000;
            const cacheNodes = this.getAll();
            let nodeVersions = [];
            for (const nodeName of Object.keys(cacheNodes)) {
                const nodeData = cacheNodes[nodeName];
                if (!nodeData.needRefresh(nextRefreshInterval)) {
                    continue;
                }
                nodeVersions.push({
                    node: nodeName,
                    v: nodeData.getVersionCode(),
                });
                if (nodeVersions.length >= REFRESH_NODE_COUNT) {
                    await this.refreshMultiKeys(nodeVersions);
                    nodeVersions = [];
                }
            }
            if (nodeVersions.length > 0) {
                this.refreshMultiKeys(nodeVersions);
            }
        }
    }
    async refreshMultiKeys(keyVersionMap) {
        let nodeDataList = [];
        try {
            nodeDataList = await this.bconfigClient.multiGetRemote(keyVersionMap);
        }
        catch (error) {
            this.logger.warn('[tcc-general] multiGetRemote get error');
        }
        nodeDataList.forEach((nodeData) => {
            const nodeName = nodeData.nodeName;
            const cachedNodeData = this._get(nodeName);
            if (!cachedNodeData) {
                return;
            }
            const lastErr = nodeData.err;
            if (lastErr && lastErr.code === error_1.TccErrorCode.errNotModified) {
                cachedNodeData.renew();
            }
            if (!lastErr || lastErr.code === error_1.TccErrorCode.errKeyNotFound) {
                this.set(nodeName, new node_1.default({
                    nodeName,
                    data: nodeData.nodeData,
                    cacheTime: cachedNodeData.getCacheTime(),
                    lastErr,
                }));
            }
        });
    }
}
exports.default = CacheNode;
