// @byted-service/tcc-general/lib/consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sleep = exports.GetType = exports.MultiResCode = exports.DEFAULT_FETCH_OPTIONS = exports.HttpStatusCode = exports.COMM_HEADERS = exports.NodeModel = exports.DEFAULT_CLUSTER_NAME = exports.TCC_BYTEDANCE_CONFIG_PROXY_PSM = exports.TCC_B_CONFIG_PROXY_PSM = exports.DEFAULT_M_GET_TIMEOUT = exports.ResErrorCode = void 0;
const env_1 = require("@byted-service/env");
var ResErrorCode;
(function (ResErrorCode) {
    ResErrorCode[ResErrorCode["respErrCodeOK"] = 0] = "respErrCodeOK";
    ResErrorCode[ResErrorCode["respErrCodeNotModified"] = 304] = "respErrCodeNotModified";
    ResErrorCode[ResErrorCode["respErrCodeNodeNotFound"] = 100] = "respErrCodeNodeNotFound";
})(ResErrorCode = exports.ResErrorCode || (exports.ResErrorCode = {}));
exports.DEFAULT_M_GET_TIMEOUT = (0, env_1.isInTCE)() ? 500 : 5000; // default get timeout of request to remote server
exports.TCC_B_CONFIG_PROXY_PSM = 'tcc.bconfig.proxy';
exports.TCC_BYTEDANCE_CONFIG_PROXY_PSM = 'bytedance.bconfig.proxy';
exports.DEFAULT_CLUSTER_NAME = 'default';
var NodeModel;
(function (NodeModel) {
    NodeModel[NodeModel["normal"] = 1] = "normal";
    NodeModel[NodeModel["group"] = 2] = "group";
})(NodeModel = exports.NodeModel || (exports.NodeModel = {}));
exports.COMM_HEADERS = {
    'X-Pod': (0, env_1.getPodName)(),
    'X-Psm': (0, env_1.getPSM)(),
    'X-Cluster': (0, env_1.getCluster)(),
    'X-Env': (0, env_1.getEnv)(),
};
var HttpStatusCode;
(function (HttpStatusCode) {
    HttpStatusCode[HttpStatusCode["statusNotModified"] = 304] = "statusNotModified";
    HttpStatusCode[HttpStatusCode["statusBadRequest"] = 400] = "statusBadRequest";
    HttpStatusCode[HttpStatusCode["statusOK"] = 200] = "statusOK";
    HttpStatusCode[HttpStatusCode["respErrCodeNodeNotFound"] = 100] = "respErrCodeNodeNotFound";
})(HttpStatusCode = exports.HttpStatusCode || (exports.HttpStatusCode = {}));
exports.DEFAULT_FETCH_OPTIONS = {
    disableMetrics: true,
    disableLogger: true,
    disableMesh: true,
    headers: exports.COMM_HEADERS,
};
exports.MultiResCode = {
    respErrCodeOK: undefined,
    respErrCodeNotModified: 304,
    respErrCodeNodeNotFound: 100,
};
var GetType;
(function (GetType) {
    GetType["getTypeCache"] = "cache";
    GetType["getTypeStaleCache"] = "stale_cache";
    GetType["getTypeServer"] = "server";
})(GetType = exports.GetType || (exports.GetType = {}));
const sleep = (ms) => {
    return new Promise((resolve) => {
        setTimeout(resolve, ms).unref();
    });
};
exports.sleep = sleep;
