// @byted-service/tcc-general/lib/error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TccErrorCode = void 0;
var TccErrorCode;
(function (TccErrorCode) {
    TccErrorCode[TccErrorCode["errKeyNotFound"] = 100] = "errKeyNotFound";
    TccErrorCode[TccErrorCode["errRemoteFailed"] = 500] = "errRemoteFailed";
    TccErrorCode[TccErrorCode["errNotModified"] = 304] = "errNotModified";
    TccErrorCode[TccErrorCode["errOther"] = 0] = "errOther";
    TccErrorCode[TccErrorCode["errInternal"] = 1] = "errInternal";
    TccErrorCode[TccErrorCode["errInvalidNodeMode"] = 2] = "errInvalidNodeMode";
    TccErrorCode[TccErrorCode["errGetException"] = 3] = "errGetException";
})(TccErrorCode = exports.TccErrorCode || (exports.TccErrorCode = {}));
const TCC_ERROR_MSG = {
    [TccErrorCode.errKeyNotFound]: '[tcc-general] key not found',
    [TccErrorCode.errRemoteFailed]: '[tcc-general] remote return failed',
    [TccErrorCode.errNotModified]: '[tcc-general] not modified',
    [TccErrorCode.errInternal]: '[tcc-general] internal error',
    [TccErrorCode.errInvalidNodeMode]: '[tcc-general] node mode not match',
    [TccErrorCode.errGetException]: '[tcc-general] get from remote failed',
};
class TCCError extends Error {
    constructor(code, msg) {
        super(msg ? msg : TCC_ERROR_MSG[code]);
        this.code = code;
    }
}
exports.default = TCCError;
