// @byted-service/tcc-general/lib/node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const error_1 = tslib_1.__importStar(require("./error"));
class Node {
    constructor(options) {
        this.nodeName = options.nodeName;
        this.data = options.data;
        this.lastErr = options.lastErr;
        this.cacheTime = options.cacheTime;
        this.expiredAt = Date.now() + this.cacheTime;
        this.lastAccessAt = Date.now();
    }
    getValue(key, mode) {
        this.lastAccessAt = Date.now();
        if (this.lastErr) {
            throw this.lastErr;
        }
        if (!this.data) {
            throw new error_1.default(error_1.TccErrorCode.errInternal);
        }
        if (this.data.m !== mode) {
            throw new error_1.default(error_1.TccErrorCode.errInvalidNodeMode);
        }
        if (this.data.kv[key] === undefined) {
            throw new error_1.default(error_1.TccErrorCode.errKeyNotFound);
        }
        return {
            data: this.data.kv[key],
            versionCode: this.data.v,
        };
    }
    getVersionCode() {
        return this.data ? this.data.v : '';
    }
    getCacheTime() {
        return this.cacheTime;
    }
    expired() {
        return Date.now() > this.expiredAt;
    }
    renew() {
        this.expiredAt = Date.now() + this.cacheTime;
    }
    needRefresh(refreshInterval) {
        //  last access at 1.5*cache_time ago, no need refresh
        if (this.lastAccessAt + 1.5 * this.cacheTime < Date.now()) {
            return false;
        }
        // expired before next round refresh, refresh it
        if (this.expiredAt < Date.now() + refreshInterval) {
            return true;
        }
        // if this refresh can renew 2/3 cache time even more, refresh it
        if (this.expiredAt - Date.now() < this.cacheTime / 3) {
            return true;
        }
        /* istanbul ignore next */
        return false;
    }
}
exports.default = Node;
