// @byted-service/tcc-general/lib/tccGeneral.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const env_1 = require("@byted-service/env");
const cacheNode_1 = tslib_1.__importDefault(require("./cacheNode"));
const logger_1 = require("./logger");
const consts_1 = require("./consts");
const emiter_1 = tslib_1.__importDefault(require("./emiter"));
const error_1 = tslib_1.__importStar(require("./error"));
const DEFAULT_CACHE_TIME = 60 * 1000;
const NORMAL_NODE_DEFAULT_KEY = '__default__';
const DEAULT_GET_TIMEOUT = (0, env_1.isInTCE)() ? 200 : 5000;
const AT_LEAST_CACHE_TIME = 30 * 1000;
const DEFAULT_CLUSTER = 'default';
const CLIENT_MANAGER = {};
class TccGeneral {
    /**
     * 创建实例
     * @param options
     */
    static create(options) {
        options.cluster = options.cluster || DEFAULT_CLUSTER;
        const clientKey = `${options.serviceName}-${options.cluster}`;
        if (CLIENT_MANAGER[clientKey]) {
            return CLIENT_MANAGER[clientKey];
        }
        CLIENT_MANAGER[clientKey] = new TccGeneral(options);
        return CLIENT_MANAGER[clientKey];
    }
    constructor(options) {
        this.serviceName = options.serviceName;
        this.cluster = options.cluster || DEFAULT_CLUSTER;
        this.getTimeout = options.getTimeout || DEAULT_GET_TIMEOUT;
        this.logger = (0, logger_1.getLogger)(options.disableLog);
        this.cache = cacheNode_1.default.create(this.cluster, {
            getTimeout: options.getTimeout,
            logger: this.logger,
        });
        this.metricsEmiter = new emiter_1.default(this.serviceName, this.cluster);
        if (typeof options.cacheTime === 'number' &&
            options.cacheTime < AT_LEAST_CACHE_TIME) {
            this.logger.warn(`[tcc-general] cache timeout you set is ${options.cacheTime}, can not below ${AT_LEAST_CACHE_TIME}, so use ${AT_LEAST_CACHE_TIME} ms`);
            options.cacheTime = AT_LEAST_CACHE_TIME;
        }
        this.cacheTime = options.cacheTime || DEFAULT_CACHE_TIME;
    }
    /**
     * 获取普通配置节点
     * @param nodeName
     */
    async get(nodeName) {
        if (!nodeName) {
            throw new error_1.default(error_1.TccErrorCode.errOther, `[tcc-general] nodeName is empty string`);
        }
        return await this.getWithCache(nodeName, NORMAL_NODE_DEFAULT_KEY, consts_1.NodeModel.normal);
    }
    /**
     * 获取组配置节点
     * @param nodeName
     * @param key
     */
    async getGroupKey(nodeName, key) {
        if (!nodeName) {
            throw new error_1.default(error_1.TccErrorCode.errOther, `[tcc-general] nodeName is empty string`);
        }
        if (!key) {
            throw new error_1.default(error_1.TccErrorCode.errOther, `[tcc-general] key is empty string`);
        }
        return await this.getWithCache(nodeName, key, consts_1.NodeModel.group);
    }
    async getWithCache(nodeName, key, mode) {
        const realNodeName = `${this.serviceName}/${nodeName}`;
        const { configNodeData, getType } = await new Promise(async (reslve, reject) => {
            const tid = setTimeout(() => {
                reject(new error_1.default(error_1.TccErrorCode.errOther, '[tcc-general] get timeout'));
                clearTimeout(tid);
            }, this.getTimeout);
            try {
                const res = await this.cache.get(realNodeName, this.cacheTime);
                reslve(res);
            }
            catch (error) {
                /* istanbul ignore next */
                reject(error);
            }
            finally {
                clearTimeout(tid);
            }
        });
        try {
            return configNodeData.getValue(key, mode);
        }
        catch (error) {
            this.metricsEmiter.emit(error, nodeName, getType);
            throw error;
        }
    }
}
exports.default = TccGeneral;
