// @universe-design/react/es/hooks/useSize.js
import React, { useCallback, useMemo } from 'react';
import { isFunction } from '../util';
import { useMountedState } from './useMountedState';
import { useIsomorphicLayoutEffect as useLayoutEffect } from './useIsomorphicLayoutEffect';
import { getSize, useResizeObserver } from '@universe-design/react-utilities';
const defaults = {
  rect: {
    width: 0,
    height: 0,
    top: 0,
    left: 0
  }
};
export function getPositionAndSize(target, entry) {
  const size = getSize(target, entry);
  const {
    left,
    top,
    bottom,
    right,
    x,
    y
  } = target.getBoundingClientRect();
  return {
    left,
    top,
    bottom,
    right,
    x,
    y,
    width: size.width,
    height: size.height
  };
}
export function useSize() {
  let {
    onChange,
    resizeObserverDisabled = false
  } = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  const [rect, setRect] = React.useState(defaults.rect);
  const isMounted = useMountedState();
  const handleResize = useCallback(_size => {
    if (!isMounted()) {
      return;
    }
    setRect(_rect => {
      if (_size.width !== _rect.width || _size.height !== _rect.height) {
        return {
          ..._rect,
          width: _size.width,
          height: _size.height
        };
      }
      return _rect;
    });
  }, [isMounted]);
  const handleReLayout = useCallback(_position => {
    if (!isMounted()) {
      return;
    }
    setRect(_rect => {
      if (_position.left !== _rect.left || _position.top !== _rect.top) {
        return {
          ..._rect,
          left: _position.left,
          top: _position.top
        };
      }
      return _rect;
    });
  }, [isMounted]);
  const ref = useResizeObserver((entry, observer) => {
    const {
      width,
      height,
      left,
      top
    } = getPositionAndSize(entry.target, entry);
    handleResize({
      width,
      height
    });
    handleReLayout({
      left,
      top
    });
    isFunction(onChange) && onChange(entry, observer);
  }, {
    disabled: resizeObserverDisabled
  });
  const updateRect = useCallback(() => {
    if (ref.current) {
      const {
        width,
        height,
        left,
        top
      } = getPositionAndSize(ref.current);
      handleResize({
        width,
        height
      });
      handleReLayout({
        left,
        top
      });
    }
  }, [handleReLayout, handleResize, ref]);
  useLayoutEffect(() => {
    // 1. 对 inline non-replaced element 进行一个简易的兜底
    if (ref.current) {
      const {
        width,
        height,
        left,
        top
      } = getPositionAndSize(ref.current);
      handleResize({
        width,
        height
      });
      handleReLayout({
        left,
        top
      });
    }
  }, [handleReLayout, handleResize, ref]);
  return [ref, useMemo(() => ({
    ...rect,
    x: rect.left,
    y: rect.top,
    right: rect.width + rect.left,
    bottom: rect.height + rect.top
  }), [rect]), {
    handleResize,
    handleReLayout,
    updateRect
  }];
}
