// @universe-design/react/es/hooks/useScrollStatus.js
import { useCallback } from 'react';
import { useImmer } from 'use-immer';
import { getThunkElement } from '../util/getThunkElement';

/**
 * 检测滚动容器的滚动状态
 */
export function useScrollStatus() {
  let params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  const {
    disableHorizontal = false,
    disableVertical = false
  } = params;
  const [overflow, updateOverflow] = useImmer({
    t: false,
    b: false,
    l: false,
    r: false
  });
  const checkOverflow = useCallback(target => {
    const targetEl = getThunkElement(target);
    if (!targetEl) {
      return;
    }
    const {
      scrollTop,
      scrollHeight,
      clientHeight,
      scrollLeft,
      scrollWidth,
      clientWidth
    } = targetEl;
    const targetInset = {
      top: 0,
      bottom: clientHeight,
      left: 0,
      right: clientWidth
    };
    const childInset = {
      top: -scrollTop,
      bottom: -scrollTop + scrollHeight,
      left: -scrollLeft,
      right: -scrollLeft + scrollWidth
    };
    updateOverflow(prevOverflow => ({
      t: disableVertical ? prevOverflow.t : targetInset.top > childInset.top,
      b: disableVertical ? prevOverflow.b : targetInset.bottom < childInset.bottom,
      l: disableHorizontal ? prevOverflow.l : targetInset.left > childInset.left,
      r: disableHorizontal ? prevOverflow.r : targetInset.right < childInset.right
    }));
  }, [disableHorizontal, disableVertical, updateOverflow]);
  return {
    checkOverflow,
    overflow
  };
}
