// @universe-design/react/src/slider/stories/homeless.stories.tsx
/* eslint-disable @typescript-eslint/no-magic-numbers */
import React, { useState } from 'react';
import { Slider, Switch, Divider, Button } from '../..';

import './index.less';

export default {
  title: 'Components/Slider/临时示例',
  component: Slider,
};

const marks = {
  0: '0°C',
  26: '26°C',
  37: '37°C',
  100: {
    style: {
      color: '#f50',
    },
    label: <strong style={{ color: 'red' }}>100°C</strong>,
  },
};

// eslint-disable-next-line max-len
// const marks2 = { 0: '0', 10: '10', 20: '20', 30: '30', 40: '40', 50: '50', 60: '60', 70: '70', 80: '80', 90: '90', 100: '100' };

export const DifferentStates = () => {
  const [disabled, setDisabled] = useState<boolean>();
  return (
    <div>
      <div className="sw">
        <span className="l">禁用: </span>
        <Switch checked={disabled} onChange={setDisabled} />
      </div>
      <Divider text="基本" />
      <div style={{ padding: 40 }}>
        <Slider min={10} max={40} defaultValue={15} disabled={disabled} />
        <Slider defaultValue={200} disabled={disabled} />
        <Slider defaultValue={-200} disabled={disabled} />
        <Slider defaultValue={20} disabled={disabled} />
        <Slider step={null} disabled={disabled} />
        <Slider defaultValue={20} disabled={disabled} />
        <Slider step={10} defaultValue={20} dots disabled={disabled} />
        <Slider tooltipVisible tooltipPlacement="bottom" step={10} defaultValue={20} dots disabled={disabled} />
      </div>
      <Divider text="带标签的滑块" />
      <div style={{ padding: 40 }}>
        <Slider step={4} marks={marks} defaultValue={37} disabled={disabled} />
        <Slider step={4} dots marks={marks} defaultValue={37} disabled={disabled} />
        <Slider step={10} dots marks={marks} defaultValue={37} disabled={disabled} />
        <Slider step={null} marks={marks} defaultValue={37} disabled={disabled} />
        <Slider step={null} marks={{ 50: '50' }} defaultValue={37} disabled={disabled} />
      </div>
      <Divider text="自定义 tooltip 内容" />
      <div style={{ padding: 40 }}>
        <Slider step={4} marks={marks} defaultValue={37} tipFormatter={v => `${v}%`} disabled={disabled} />
      </div>
    </div>
  );
};

export const RangeStates = () => {
  const [disabled, setDisabled] = useState<boolean>();
  return (
    <div>
      <div className="sw">
        <span className="l">禁用: </span>
        <Switch checked={disabled} onChange={setDisabled} />
      </div>
      <Divider text="基本" />
      <div style={{ padding: 40 }}>
        <Slider range min={10} max={40} defaultValue={[15, 25]} disabled={disabled} />
        <Slider range defaultValue={[20, 40]} disabled={disabled} />
        <Slider range defaultValue={[20, 40]} disabled={disabled} />
        <Slider range step={10} defaultValue={[20, 40]} dots disabled={disabled} />
        <Slider
          tooltipVisible
          tooltipPlacement="bottom"
          range
          step={10}
          defaultValue={[20, 40]}
          dots
          disabled={disabled}
        />
      </div>
      <Divider text="带标签的滑块" />
      <div style={{ padding: 40 }}>
        <Slider range step={4} marks={marks} defaultValue={[20, 40]} disabled={disabled} />
        <Slider range step={10} dots marks={marks} defaultValue={[20, 40]} disabled={disabled} />
        <Slider range step={null} marks={marks} defaultValue={[20, 40]} disabled={disabled} />
        <Slider range step={null} marks={{ 50: '50' }} defaultValue={[20, 40]} disabled={disabled} />
      </div>
      <Divider text="自定义 tooltip 内容" />
      <div style={{ padding: 40 }}>
        <Slider range step={4} marks={marks} defaultValue={[20, 40]} tipFormatter={v => `${v}%`} disabled={disabled} />
        <Slider
          range
          included
          step={4}
          marks={marks}
          defaultValue={[20, 40]}
          tipFormatter={v => `${v}%`}
          disabled={disabled}
        />
      </div>
    </div>
  );
};

export const VerticalStates = () => {
  const [disabled, setDisabled] = useState<boolean>(false);
  return (
    <div className="ud-slider-demo">
      <div className="sw">
        <span className="l">禁用: </span>
        <Switch checked={disabled} onChange={setDisabled} />
      </div>
      <Divider text="基本" />
      <div className="block">
        <Slider vertical min={10} max={40} defaultValue={15} disabled={disabled} />
        <Slider vertical defaultValue={200} disabled={disabled} />
        <Slider vertical defaultValue={-200} disabled={disabled} />
        <Slider vertical defaultValue={20} disabled={disabled} />
      </div>
      <div className="block">
        <Slider vertical step={10} dots defaultValue={20} disabled={disabled} />
      </div>
      <div className="block">
        <Slider vertical range max={40} min={10} defaultValue={[15, 25]} disabled={disabled} />
        <Slider vertical range defaultValue={[20, 40]} disabled={disabled} />
      </div>
      <div className="block">
        <Slider vertical range step={10} defaultValue={[20, 40]} dots disabled={disabled} />
      </div>
      <div className="block">
        <Slider tooltipVisible vertical range step={10} defaultValue={[20, 40]} dots disabled={disabled} />
      </div>
      <Divider text="带标签的滑块" />
      <div className="block">
        <Slider vertical range step={4} marks={marks} defaultValue={[20, 40]} disabled={disabled} />
      </div>
      <div className="block">
        <Slider vertical range step={10} dots marks={marks} defaultValue={[20, 40]} disabled={disabled} />
      </div>
      <div className="block">
        <Slider vertical range step={null} marks={marks} defaultValue={[20, 40]} disabled={disabled} />
      </div>
      <div className="block">
        <Slider vertical range step={null} marks={{ 50: '50' }} defaultValue={[20, 40]} disabled={disabled} />
      </div>
    </div>
  );
};

export const ControlledRange = () => {
  const [value, setValue] = useState<[number, number]>([20, 40]);
  const onChange = (v: [number, number]) => {
    console.log('onChange', v);
    setValue(v);
  };
  return <Slider range value={value} onChange={onChange} />;
};

export const Controlled = () => {
  const [value, setValue] = useState<number>(20);
  const onChange = (v: number) => {
    console.log('onChange', v);
    setValue(v);
  };
  return <Slider value={value} onChange={onChange} />;
};

export const ChangeDirection = () => {
  const [vertical, setVertical] = useState(false);
  return (
    <div style={{ height: 200 }}>
      <Slider
        style={{
          width: 400,
        }}
        vertical={vertical}
        defaultValue={30}
      />
      <Button onClick={() => setVertical(!vertical)}>切换方向</Button>
    </div>
  );
};
