// @universe-design/react/src/slider/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import CBasic from '../demo/basic';
import CDisabled from '../demo/disabled';
import CEvent from '../demo/event';
import CMark from '../demo/mark';
import CRange from '../demo/range';
import CShowTooltip from '../demo/show-tooltip';
import CTipFormatter from '../demo/tip-formatter';
import CVertical from '../demo/vertical';
if (!(CBasic as Story).storyName) {
  (CBasic as Story).storyName = '0. 基础用法';
  (CBasic as Story).parameters = {
    docs: {
      description: {
        story: '简单的用法。\n\n',
      },
    },
  };
}
if (!(CDisabled as Story).storyName) {
  (CDisabled as Story).storyName = '1. 禁用状态';
  (CDisabled as Story).parameters = {
    docs: {
      description: {
        story: '设置`disabled`为`true`时，滑块处于不可用状态。\n\n',
      },
    },
  };
}
if (!(CEvent as Story).storyName) {
  (CEvent as Story).storyName = '5. 监听改变';
  (CEvent as Story).parameters = {
    docs: {
      description: {
        story: '当值发生变化时，会触发`onChange`事件。\n\n',
      },
    },
  };
}
if (!(CMark as Story).storyName) {
  (CMark as Story).storyName = '6. 添加标签';
  (CMark as Story).parameters = {
    docs: {
      description: {
        story:
          '通过`marks`设置标签文本。当`included`为`false`时，表明不同标记间为并列关系。当`step`为 null 时，可选值仅有`marks`标出来的部分。\n\n',
      },
    },
  };
}
if (!(CRange as Story).storyName) {
  (CRange as Story).storyName = '2. 双滑块';
  (CRange as Story).parameters = {
    docs: {
      description: {
        story: '设置`range`为`true`时，渲染为双滑块。\n\n',
      },
    },
  };
}
if (!(CShowTooltip as Story).storyName) {
  (CShowTooltip as Story).storyName = '8. 控制 Tooltip 的显示';
  (CShowTooltip as Story).parameters = {
    docs: {
      description: {
        story:
          '当`tooltipVisible`为`true`时，将始终显示 ToolTip。当`tooltipVisible`为`false`时，将始终不显示 ToolTip。\n\n',
      },
    },
  };
}
if (!(CTipFormatter as Story).storyName) {
  (CTipFormatter as Story).storyName = '4. 自定义提示';
  (CTipFormatter as Story).parameters = {
    docs: {
      description: {
        story: '使用`tipFormatter`可以格式化`Tooltip`的内容。`tipFormatter`为 null 时隐藏`Tooltip`。\n\n',
      },
    },
  };
}
if (!(CVertical as Story).storyName) {
  (CVertical as Story).storyName = '7. 竖直滑动条';
  (CVertical as Story).parameters = {
    docs: {
      description: {
        story: '通过设置`vertical`为`true`开启竖直模式。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Slider/官网示例',
  tags: ['autodocs'],
};

export { CBasic, CDisabled, CEvent, CMark, CRange, CShowTooltip, CTipFormatter, CVertical };
