// @universe-design/react/src/drawer/Drawer.tsx
/* eslint-disable @typescript-eslint/no-magic-numbers */
import React, { useCallback } from 'react';
import stylex from '@universe-design/stylex';
import CloseOutlined from '@universe-design/icons-react/CloseOutlined';
import isNil from 'lodash/isNil';
import clsx from 'clsx';
import { Transition } from 'react-transition-group';
import Portal from '../portal';
import { Button } from '../button';
import { useCssScope } from '../config-provider/hooks';
import { FocusLock } from '../shared';
import type { DefaultSizeType, FocusableElement } from '../util';
import { getDefaultSizeAbbr, isFunction } from '../util';
import { useAnimatedVisible } from '../shared/components/transition/useAnimatedVisible';
import { FadeTransition } from '../shared/components/transition/FadeTransition';
import type { SlideDirection } from '../shared/components/transition/SlideTransition';
import { SlideTransition } from '../shared/components/transition/SlideTransition';
import type { ThunkElement } from '../util/getThunkElement';
import { cssTypographyBody0, cssTypographyTitle4 } from '../stylex';

import './style';

/**
 * @public
 */
export type EventType = React.KeyboardEvent<HTMLDivElement> | React.MouseEvent<HTMLDivElement | HTMLButtonElement>;

export type PlacementType = 'top' | 'right' | 'bottom' | 'left';

export interface DrawerProps extends React.PropsWithChildren<unknown> {
  prefixCls?: string;
  /**
   * 抽屉外层容器的 `className` 属性透传
   */
  className?: string;
  /**
   * 抽屉外层容器的 `style` 属性透传
   */
  style?: React.CSSProperties;
  /**
   * 抽屉内容的 `style` 属性透传
   */
  bodyStyle?: React.CSSProperties;
  /**
   * 抽屉弹出层的 `style` 属性透传
   */
  drawerStyle?: React.CSSProperties;
  /**
   * 抽屉页脚的 `style` 属性透传
   */
  footerStyle?: React.CSSProperties;
  /**
   * 抽屉头部的 `style` 属性透传
   */
  headerStyle?: React.CSSProperties;
  /**
   * 抽屉遮罩的 `style` 属性透传
   */
  maskStyle?: React.CSSProperties;
  /**
   * 是否显示遮罩层
   * @defaultValue true
   */
  mask?: boolean;
  /**
   * 点击蒙层是否允许关闭
   * @defaultValue true
   */
  maskClosable?: boolean;
  /**
   * 自定义关闭图标
   * @defaultValue `<CloseOutlined />`
   */
  closeIcon?: React.ReactNode;
  /**
   * 自定义页脚
   */
  footer?: React.ReactNode;
  /**
   * 自定义标题
   */
  title?: React.ReactNode;
  /**
   * 自定义头部扩展部分内容
   */
  headerExtra?: React.ReactNode;
  /**
   * 自定义副标题
   */
  subTitle?: React.ReactNode;
  /**
   * 自定义高度
   */
  height?: number | string;
  /**
   * 自定义宽度
   */
  width?: number | string;
  /**
   * 用于设置抽屉内容部分的 `padding` 样式
   * @defaultValue `medium`
   */
  bodyPaddingSize?: 'small' | 'medium' | number;
  /**
   * 切换抽屉时动画结束后的回调
   */
  afterVisibleChange?: (visible: boolean) => void;
  /**
   * 是否可见
   * @defaultValue false
   */
  visible?: boolean;
  /**
   * 设置 `z-index`
   */
  zIndex?: number;
  /**
   * @deprecated
   * use showCloseIcon instead
   */
  closable?: boolean;
  /**
   * 是否显示右上角的关闭按钮
   * @defaultValue false
   */
  showCloseIcon?: boolean;
  /**
   * 关闭时销毁抽屉里的子元素
   * @defaultValue false
   */
  destroyOnClose?: boolean;
  /**
   * 点击遮罩层或右上角叉或取消按钮的回调
   */
  onClose?: (e: EventType) => void;
  /**
   * 预渲染抽屉内元素
   * @defaultValue false
   */
  forceRender?: boolean;
  /**
   * 指定从抽屉挂载的 HTML 节点, `false` 为挂载在当前 DOM
   */
  getContainer?: string | HTMLElement | (() => HTMLElement | null) | false;
  /**
   * 是否支持键盘 Esc 关闭
   */
  keyboard?: boolean;
  /**
   * 抽屉的方向
   * @defaultValue `right`
   */
  placement?: PlacementType;
  /**
   * 抽屉的尺寸，仅 `placement` 为 `left` `right` 时生效
   * @defaultValue `small`
   */
  size?: DefaultSizeType;
  /**
   * 进入 drawer 时的初始 focus 的元素, 默认为 content
   */
  initialFocusRef?: ThunkElement<FocusableElement>;
  /**
   * drawer 关闭后恢复 focus 的元素
   */
  finalFocusRef?: ThunkElement<FocusableElement>;
  /**
   * 开启🤔关闭 focusLock
   * @defaultValue false
   */
  enableFocusLock?: boolean;
  /**
   * 关闭对话框后是否恢复 focus 到打开它的元素
   * @defaultValue false
   */
  enableRestoreFocus?: boolean;
}

// TODO: 1. warn 换 utils 内的 2.考虑放到 utils 下
function compatibleDeprecatedProps<T, K extends keyof T>(props: T, key: K, deprecatedKey: K) {
  if (!isNil(props[key])) {
    return props[key];
  }

  if (!isNil(props[deprecatedKey])) {
    console.warn(`"${String(deprecatedKey)}" is deprecated, please use "${String(key)}" instead.`);
    return props[deprecatedKey];
  }
}

const toSlideDirectionMap: Record<PlacementType, SlideDirection> = {
  left: 'right',
  right: 'left',
  bottom: 'up',
  top: 'down',
};

const useDrawerCssScope = () => useCssScope('drawer');

const getNumericalSize = (size: DefaultSizeType) => {
  switch (size) {
    case 'large':
      return 680;
    case 'medium':
      return 480;
    case 'small':
      return 350;
    default:
      throw new Error(`Size \`${size}\` is not support.`);
  }
};

export const Drawer: React.FC<DrawerProps> = props => {
  const {
    visible = false,
    placement = 'right',
    size = 'small',
    width = getNumericalSize(size),
    height = getNumericalSize(size),
    style: wrapperStyle,
    maskStyle,
    mask = true,
    maskClosable = true,
    keyboard = true,
    destroyOnClose = false,
    className: wrapperClassName,
    bodyPaddingSize = 'medium',
    prefixCls,
    zIndex,
    forceRender,
    getContainer,
    afterVisibleChange,
    onClose,
    footer,
    children,
    initialFocusRef,
    finalFocusRef,
    enableFocusLock = false,
    enableRestoreFocus = false,
  } = props;

  const showCloseIcon = compatibleDeprecatedProps(props, 'showCloseIcon', 'closable');

  const defaultScope = useDrawerCssScope();
  const cssScope = prefixCls || defaultScope();

  const timeout = React.useMemo(
    () => ({
      appear: 300,
      enter: 300,
      exit: 200,
    }),
    [],
  );

  const { transitionGroupProps, transitionGroupPropsWithoutListener, styleDisplay, animatedVisible, noopRef } =
    useAnimatedVisible({
      destroyOnClose,
      forceRender,
      visible,
      onAfterClose: React.useCallback(
        () => isFunction(afterVisibleChange) && afterVisibleChange(false),
        [afterVisibleChange],
      ),
      onAfterOpen: React.useCallback(
        () => isFunction(afterVisibleChange) && afterVisibleChange(true),
        [afterVisibleChange],
      ),
    });

  const slideExtraProps = React.useMemo(() => {
    const direction = toSlideDirectionMap[placement];
    return {
      direction,
      size: (['left', 'right'].includes(direction) ? width : height) ?? 0,
    };
  }, [height, placement, width]);

  const renderCloseIcon = () => {
    const { closeIcon = <Button icon={<CloseOutlined />} size="medium" type="icon" /> } = props;
    return (
      <div className={`${cssScope}__close`} onClick={onClose}>
        {closeIcon}
      </div>
    );
  };

  const renderHeader = () => {
    const { title, subTitle, headerExtra, headerStyle } = props;
    // compatible other DS's props
    if (!title && !showCloseIcon) {
      return null;
    }
    return (
      <div className={clsx(`${cssScope}__header`)} style={headerStyle}>
        <div className={`${cssScope}__header__content`}>
          {(title || subTitle) && (
            <div className={clsx(`${cssScope}__header__title`, stylex(cssTypographyTitle4))}>{title}</div>
          )}
          {showCloseIcon && renderCloseIcon()}
        </div>
        <div className={`${cssScope}__header__extra`}>
          {subTitle && <p className={clsx(`${cssScope}__header__subTitle`, stylex(cssTypographyBody0))}>{subTitle}</p>}
          {headerExtra ? headerExtra : null}
        </div>
      </div>
    );
  };

  const renderFooter = () => {
    const { footerStyle } = props;
    if (!footer) {
      return null;
    }

    const footerClassName = `${cssScope}__footer`;
    return (
      <div className={footerClassName} style={footerStyle}>
        {footer}
      </div>
    );
  };

  const onMaskClick = useCallback(
    e => {
      if (!maskClosable) {
        return;
      }
      if (onClose) {
        onClose(e);
      }
    },
    [maskClosable, onClose],
  );

  const maskRef = React.useRef<HTMLDivElement>(null);

  const renderMask = () => {
    if (!mask) {
      return;
    }

    const mergeMaskStyle: React.CSSProperties = {
      ...maskStyle,
      ...styleDisplay,
    };
    if (zIndex || zIndex === 0) {
      mergeMaskStyle.zIndex = zIndex;
    }
    return (
      <FadeTransition timeout={timeout} appear nodeRef={maskRef} {...transitionGroupPropsWithoutListener}>
        <div ref={maskRef} className={`${cssScope}__mask`} style={mergeMaskStyle} onClick={onMaskClick} tabIndex={-1} />
      </FadeTransition>
    );
  };

  const onKeyDown = useCallback(
    (e: React.KeyboardEvent) => {
      if (!keyboard) {
        return;
      }
      // ESC

      if (e.keyCode === 27) {
        e.stopPropagation();
        if (onClose) {
          onClose(e as any);
        }
      }
    },
    [onClose, keyboard],
  );

  const dialogRef = React.useRef<HTMLDivElement>(null);
  const mergedInitialFocusRef = initialFocusRef === null ? null : initialFocusRef ?? dialogRef;

  const renderContent = () => {
    const { bodyStyle: bodyStyleProps, drawerStyle } = props;
    const contentStyle: React.CSSProperties = {
      ...drawerStyle,
      ...styleDisplay,
      ...(height && ['top', 'bottom'].includes(placement) ? { height } : {}),
      ...(width && ['left', 'right'].includes(placement) ? { width } : {}),
      zIndex,
    };

    const bodyStyle: React.CSSProperties = {
      ...bodyStyleProps,
      ...(typeof bodyPaddingSize === 'number' ? { padding: `${bodyPaddingSize}px` } : {}),
    };
    const bodyPaddingKeywords = typeof bodyPaddingSize !== 'number' ? getDefaultSizeAbbr(bodyPaddingSize) : '';

    const mergedEnableFocusLock = (() => {
      // 在非 mask 情况, 即使用户没有传 enableFocusLock, 也要关掉lock, 否则按照用户传的为准
      if (!mask && !enableFocusLock) {
        return false;
      }
      return enableFocusLock;
    })();

    return (
      <SlideTransition
        appear
        timeout={timeout}
        extraProps={slideExtraProps}
        nodeRef={dialogRef}
        {...transitionGroupPropsWithoutListener}
      >
        <FocusLock
          isDisabled={!visible}
          initialFocusRef={mergedInitialFocusRef}
          restoreFocus={enableRestoreFocus}
          finalFocusRef={finalFocusRef}
          enableFocusLock={mergedEnableFocusLock}
        >
          <div
            ref={dialogRef}
            role="dialog"
            aria-modal="true"
            tabIndex={-1}
            onKeyDown={onKeyDown}
            className={clsx(`${cssScope}__content__wrapper`)}
            style={contentStyle}
          >
            {renderHeader()}
            <div
              className={clsx(`${cssScope}__content`, {
                [`${cssScope}__content__nofooter`]: footer,
                [`${cssScope}__content__pd-${bodyPaddingKeywords}`]: bodyPaddingKeywords,
              })}
              style={bodyStyle}
            >
              {children}
            </div>
            {renderFooter()}
          </div>
        </FocusLock>
      </SlideTransition>
    );
  };

  const renderPortal = () => {
    const childNodes = (
      <div
        className={clsx(`${cssScope}__root`, {
          [`${cssScope}-${placement}`]: true,
          [`${cssScope}-open`]: animatedVisible,
        })}
        style={wrapperStyle}
      >
        {renderMask()}
        {renderContent()}
      </div>
    );
    if (getContainer === false) {
      return <div className={wrapperClassName}>{childNodes}</div>;
    }
    return (
      <Portal wrapperClassName={wrapperClassName} container={getContainer}>
        {childNodes}
      </Portal>
    );
  };

  return (
    <Transition nodeRef={noopRef} timeout={timeout} appear {...transitionGroupProps}>
      {renderPortal()}
    </Transition>
  );
};
