// @universe-design/react/src/affix/index.tsx
import type { CSSProperties } from 'react';
import React, { forwardRef, useEffect, useRef, useCallback, useImperativeHandle } from 'react';
import listScrollParents from '../shared/components/trigger/@popperjs/core/lib/dom-utils/listScrollParents';
import { useImmer } from 'use-immer';
import { useIsomorphicLayoutEffect as useLayoutEffect, useSize } from '../hooks';
import {
  useResizeObserver,
  useControlledState,
  useCloneElementRef,
  isHTMLElement,
} from '@universe-design/react-utilities';
import { useCssScope, useGetTargetContainer } from '../config-provider/hooks';

import './style/index.less';

/**
 * @public
 */
export interface AffixProps {
  /**
   * 图钉样式
   */
  style?: CSSProperties;
  /**
   * 自定义 className
   */
  className?: string;
  /**
   * 距离窗口顶部达到指定偏移量后触发
   */
  offsetTop?: number;
  /**
   * 距离窗口底部达到指定偏移量后触发
   */
  offsetBottom?: number;
  /**
   * 作用的组件
   */
  children: React.ReactNode;
  /**
   * 设置 `Affix` 需要监听其滚动事件的元素，值为一个返回对应 DOM 元素的函数
   * @defaultValue () => window
   */
  target?: () => typeof window | HTMLElement;
  /**
   * 固定状态改变时触发的回调函数
   */
  onChange?: (affixed: boolean) => void;
}

export const Affix = forwardRef<HTMLDivElement, AffixProps>((props, ref) => {
  const { style, className, target: targetInProps, offsetTop, offsetBottom, children, onChange } = props;

  const cssScope = useCssScope('affix');
  const wrapperRef = useResizeObserver<HTMLDivElement>(() => {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    measure();
  });
  const placeholderRef = useRef<HTMLDivElement>(null);
  // 第一次进页面大部分情况下滚动条都是处于 0 的位置, 所以这里可以是 false
  // 如果出现刚一进来滚动条就在非 0 的位置的情况, 通过下面的 useLayoutEffect 强刷一下
  const [fixed, setFixed] = useControlledState<boolean>(false, undefined, {
    onChange,
  });
  // style 是一个对象，用 immer 可以减少一些 rerender
  const [affixStyle, produceAffixStyle] = useImmer<CSSProperties>({});
  // 监听 children 的尺寸变化
  const [childRef, childSize] = useSize<HTMLDivElement>();

  // 获取 target container
  const ctxGetTargetContainer = useGetTargetContainer();
  const target = targetInProps || ctxGetTargetContainer;

  // 更新 affix node 的样式
  const measure = useCallback(() => {
    const targetNode = target();
    const wrapperNode = wrapperRef.current;
    if (!targetNode || !wrapperNode) {
      return;
    }

    let targetRect: DOMRect;
    if (isHTMLElement(targetNode)) {
      targetRect = targetNode.getBoundingClientRect();
    } else {
      targetRect = { top: 0, bottom: window.innerHeight } as DOMRect;
    }
    const wrapperRect = wrapperNode.getBoundingClientRect();
    const { top: targetTop, bottom: targetBottom } = targetRect;
    const { top: placeholderTop, bottom: placeholderBottom } = wrapperRect;
    const offsetTopVal = offsetTop === undefined && offsetBottom === undefined ? 0 : offsetTop;
    const fixedTop =
      offsetTopVal !== undefined && targetTop > placeholderTop - offsetTopVal ? offsetTopVal + targetTop : undefined;
    const fixedBottom =
      offsetBottom !== undefined && targetBottom < placeholderBottom + offsetBottom
        ? offsetBottom + (window.innerHeight - targetBottom)
        : undefined;
    if (fixedTop !== undefined) {
      setFixed(true);
      produceAffixStyle({
        top: fixedTop,
        width: wrapperRect.width,
        height: wrapperRect.height,
      });
    } else if (fixedBottom !== undefined) {
      setFixed(true);
      produceAffixStyle({
        bottom: fixedBottom,
        width: wrapperRect.width,
        height: wrapperRect.height,
      });
    } else {
      setFixed(false);
      produceAffixStyle({});
    }
  }, [offsetBottom, offsetTop, produceAffixStyle, setFixed, target, wrapperRef]);

  // TODO: 给 scroll 用的 lazyMeasure 方法，这块后续可能还要优化一下，加一下 throttle
  const lazyMeasure = React.useCallback(() => {
    requestAnimationFrame(() => {
      measure();
    });
  }, [measure]);

  useEffect(() => {
    const _target = target();
    _target?.addEventListener('scroll', lazyMeasure);
    _target?.addEventListener('resize', measure);
    /**
     * 有一种场景, 当 anchor 的 container 容器之上的 parents 们中的谁如果有个高度
     * 然后 anchor 的 container 的高度又大于这个 parent 的高度并且这个 parent 也允许溢出滚动的时候
     * 会出现这种情况:
     *    当 anchor 的 container 滚动到头了,
     *    但是由于 container 的 height 比它的某一个 parent 的 height 要高
     *    所以此时再继续向下滚动的时候, 就会触发这个 parent 的 scroll 事件
     *    此时页面会继续向下滚, 但是由于不会再触发 anchor 的 container 的 scroll 事件
     *    就会导致这个 anchor 的位置不对了
     *
     * 所以还需要对所有的可以滚动的父元素进行 scroll 的监听
     */
    const scrollParents = _target && listScrollParents(_target as Element);
    scrollParents?.forEach(element => {
      element?.addEventListener('scroll', lazyMeasure);
    });
    return () => {
      _target?.removeEventListener('scroll', lazyMeasure);
      _target?.removeEventListener('resize', measure);
      scrollParents?.forEach(element => {
        element?.removeEventListener('scroll', lazyMeasure);
      });
    };
  }, [lazyMeasure, measure, target]);

  // 防止第一次进来滚动条就已经在页面中间
  useLayoutEffect(() => {
    measure();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    // 这里使用同步设置状态是因为 measure 里面要用到 wrapper 的 size 去更新 affix size，wrapper size 是通过 placeholder size 去撑开的，某些情况下如果 placeholder size 没有同步设置，会导致 render cycle
    if (placeholderRef.current) {
      placeholderRef.current.style.setProperty('width', `${childSize.width}px`);
      placeholderRef.current.style.setProperty('height', `${childSize.height}px`);
    }
    measure();
  }, [measure, childSize.height, childSize.width]);

  useImperativeHandle<HTMLDivElement | null | undefined, HTMLDivElement | null | undefined>(
    ref,
    () => wrapperRef.current,
  );

  const renderPlaceholderElement = () => (fixed ? <div ref={placeholderRef} /> : undefined);

  const _children = useCloneElementRef({ children, ref: childRef });

  return (
    <div className={className} style={style} ref={wrapperRef}>
      {renderPlaceholderElement()}
      {/* 这里不 fixed 的时候就直接返回 children 是为了解决 children height: 100% 这种 case */}
      {fixed ? (
        <div className={cssScope()} style={affixStyle}>
          {_children}
        </div>
      ) : (
        _children
      )}
    </div>
  );
});
