// @lark/moment-timezone/src/core.ts

import moment from '@lark/moment-timezone-polyfill'
import { IStartDayOfWeek } from './constants'

export const supportLocales = [
  'de-DE',
  'en-US',
  'es-ES',
  'fr-FR',
  'hi-IN',
  'id-ID',
  'it-IT',
  'ja-JP',
  'ko-KR',
  'ms-MY',
  'pt-BR',
  'ru-RU',
  'th-TH',
  'vi-VN',
  'zh-CN',
  'zh-HK',
  'zh-TW',
]

let HOUR_TWELVE = false

/**
 * @description 初始化moment
 * @param { locale, localeConfig, is12Hour, timezone }
 * locale: locale的key，项目不维护
 * localeConfig: import('moment-timezone').LocaleSpecification类型，业务方可以从locales中选取适配的语言做导入
 * is12Hour: 是否是12小时制
 * timezone: 时区id，如Asia/Shanghai等
 */
export function initMoment ({
  locale,
  localeConfig,
  is12Hour = false,
  timezone,
}: {
  locale: string;
  localeConfig: moment.LocaleSpecification | moment.Locale;
  is12Hour: boolean;
  timezone?: string;
}): void {
  // 12hour or 24hour format
  HOUR_TWELVE = is12Hour
  locale = locale.replace(/(\w+)[-_](\w+)/, function (m, $1, $2) {
    return `${$1}-${$2.toUpperCase()}`
  })
  if (timezone) moment.tz.setDefault(timezone)
  moment.updateLocale(locale, localeConfig as moment.LocaleSpecification)
}

/**
 * @description 更新是否是12h制
 * @param is12Hour
 */
export function updateTimeFormatHour (is12Hour: boolean) {
  HOUR_TWELVE = is12Hour
}

export function is12HourFormat () {
  return HOUR_TWELVE
}

/**
 * @description 更新一周的开始时间
 * @param startDayOfWeek
 */
export function updateStartDayOfWeek (startDayOfWeek: IStartDayOfWeek) {
  const currentLocale = moment.locale()
  moment.updateLocale(currentLocale, startDayOfWeek)
}

/**
 * @description 获得一周的开始时间
 */
export function getStartDayOfWeek () {
  const currentLocale = moment.localeData()
  return currentLocale.firstDayOfWeek()
}

/**
 * @see 与原逻辑不一样，不再维护localtimezone的全局变量
 * @description 设置moment默认时区
 * @param timezone
 */
export function updateLocalTimezone (timezone: string) {
  moment.tz.setDefault(timezone)
}

/**
 * @see 与原逻辑不一样，不再维护localtimezone的全局变量
 * @description 没有调用initMoment的时候，返回的就是undefinded
 */
export function getLocalTimezone () {
  const timezone = moment().tz()
  if (!timezone) {
    return moment.tz.guess(true)
  }
  return timezone
}
