// @lark/moment-timezone/src/extend.ts
import moment from '@lark/moment-timezone-polyfill'
import { getLocalTimezone } from './core'

/**
 * @description 基于原来的显示时间，强制修改时区（时间戳会改掉）
 * @param momentDate 
 * @param timezone 
 */
export function forceChangeTimezone (momentDate: moment.Moment, timezone: string): moment.Moment {
  const formattedTime = momentDate.format('YYYY-MM-DD HH:mm')
  return moment.tz(formattedTime, timezone)
}

/**
 * @description 根据Unix时间戳、时区及是否全天生成, 
 * @param momentDate 
 * @param timezone 
 */
export function getMomentByUnix ({ timestamp, timezone, isAllDay = false }: { timestamp: number; timezone: string; isAllDay?: boolean; }): moment.Moment {
  if (isAllDay && timezone === 'UTC') {
    const utcDate = moment.tz(timestamp * 1000, timezone)
    return moment(utcDate.format('YYYY-MM-DD'))
  }
  return moment.tz(timestamp * 1000, timezone)
}

/**
 * @description 根据时区id，获得该时区的当前日期
 * @param timezone 
 */
export function getDateInTimezone (timezone: string): number {
  return moment().tz(timezone).date()
}

/**
 * @see 原changeMomentTimezone
 * @description 把moment对象转成default时区, 如果没有设置default时区，则返回原对象
 * @param momentDate 
 */
export function change2DefaultTimezone (momentDate: moment.Moment): moment.Moment {
  const currentTimezone = moment().tz()
  if (!currentTimezone) {
    return momentDate
  }
  return moment.tz(momentDate, currentTimezone)
}

/**
 * @see 和原逻辑不一样，不再支持传入时区
 * @description 获得default时区的当前时间对象
 */
export function getLocalNow () {
  return moment()
}

/**
 * @see 为什么使用了kk:mm这个format，kk返回的是1~24
 * @description 获得default时区的分钟数
 */
export function getLocalNowMinutes () {
  const time = getLocalNow().format('kk:mm').split(':').map(Number)
  return time[0] * 60 + time[1]
}

/**
 * @see 因为不再有localTimezone全局变量，所以实现不一样，需要注意看看
 * @description 根据unix时间戳生成default timezone的momemnt对象
 * @param unixTimestamp 
 */
export function getLocalMoment (unixTimestamp) {
  return moment(unixTimestamp * 1000)
}

/**
 * @deprecated 其实就是moment.unix方法的封装，没有必要
 * @description 返回moment对象的Unix时间戳
 * @param time 
 */
export function getLocalUnix (time: moment.Moment) {
  return time.unix()
}

/**
 * @description 获得default时区的当前时间戳，在UTC时区下当天开始时间的unix时间戳
 * 目前仅用于rrule
 */
export function getUTCTodayStartUnix () {
  const now = moment()
  return now.tz('UTC').startOf('day').unix()
}

/**
 * @description 获得当前时刻的unix时间戳
 */
export function getLocalUnixNow () {
  return getLocalNow().unix()
}

/**
 * @see 原getNewWeekStartDate
 * @description 获得当前当前日期的moment对象（重置一周开始时间后调用）
 * @param momentDate 
 */
export function getWeekStartDate (momentDate: moment.Moment) {
  const time = momentDate.format('YYYY-MM-DD')
  return moment(time)
}

/**
 * @see 由于没有localtimezone全局变量，实现不一样
 * @description Date对象转moment对象
 * @param date 
 */
export function getLocalMomentFromDate (date: Date) {
  return moment.tz(date, getLocalTimezone())
}

/**
 * @see 和接口不一样，不再支持timeFormat默认值
 * @description 从时间字符串及相应的formatter构造default时区的moment对象
 * @param dateString 
 * @param timeFormat 
 */
export function getMomentFromDateString (dateString: string, timeFormat) {
  return moment(dateString, timeFormat)
}

/**
 * @description minutesDiff实际就是某一时刻的距离当前开始时间在分钟上的事件差，通过这个更新时间
 * @param { date: moment.Moment, minutesDiff: number }
 */
export function getMomentByMinutesDiff ({ date, minutesDiff }: { date: moment.Moment, minutesDiff: number }) {
  const _hours = ~~(minutesDiff / 60)
  const _minutes = minutesDiff - _hours * 60
  return date.clone().hours(_hours).minutes(_minutes)
}
