// @lark/moment-timezone/src/format.ts
import moment from '@lark/moment-timezone-polyfill'
import { is12HourFormat, supportLocales } from './core'

export type IFormatOption = {
  locale?: typeof supportLocales[number];
  is12Hour?: boolean;
  timezone?: string;
}
const contentRegx = /\{\{([_$a-zA-Z0-9]+)\}\}/g
const delimiter = '||||'

/**
 * @description i18n模板语法
 * @param template 
 * @param context 
 */
const __t = (template: string, context?: object) => template.replace(contentRegx, (match, key) => {
  if (!context || context[key] === undefined) {
    return ''
  }
  return context[key]
})

/**
 * @description 12h增强的i18n模板语法
 * 'Ah:mm||||HH:mm', 如果是12h选择第一个，如果是24h选择第二个
 * @param template 
 * @param context 
 */
export const __T
  = (template: string, context: object = {}, is12Hour: boolean = is12HourFormat()) => {
    const templates = template.split(delimiter)
    // 没有两个语言的情况
    if (templates.length === 1) {
      return __t(templates[0], context)
    }

    if (is12Hour) {
      return __t(templates[0], context)
    }

    return __t(templates[1], context)
  }

/**
 * 根据定制模板，获得moment库
 * @param date 
 * @param template 
 * @param context 
 */
export function formatMoment (date: moment.Moment, template: string, context: object = {}, {
  locale,
  is12Hour,
  timezone,
}: IFormatOption = {}): string {
  const formatter = __T(template, context, is12Hour)
  let formatDate = date
  if (timezone) formatDate = formatDate.tz(timezone)
  if (locale) formatDate = formatDate.locale(locale)
  return formatDate.format(formatter)
}

/**
 * @description 根据定制模板，获得format函数
 * @param template 
 */
export function getMomentFormatter (template: string) {
  return function (date: moment.Moment, context: object = {}, option: IFormatOption = {}): string {
    return formatMoment(date, template, context, option)
  }
}

/**
 * @description 获得weekday缩写数组，如Su_Mo_Tu_We_Th_Fr_Sa
 * @param orderByLocale 
 */
export function getWeekdaysMin (orderByLocale = true) {
  return moment.weekdaysMin(orderByLocale)
}

/**
 * @description 获得weekdayshort缩写数组，如Sun_Mon_Tue_Wed_Thu_Fri_Sat
 * @param orderByLocale 
 */
export function getWeekdaysShort (orderByLocale = true) {
  return moment.weekdaysShort(orderByLocale)
}

/**
 * @description 获得weekdays数组，如Sunday_Monday_Tuesday_Wednesday_Thursday_Friday_Saturday
 * @param orderByLocale 
 */
export function getWeekdays (orderByLocale = true) {
  return moment.weekdays(orderByLocale)
}

/**
 * 根据时区id，获得该时区id对应时间下的 GMToffset的字符串表示 如 GMT+08:00 / GMT+8
 * @param timezone 
 * @param { padNumber, date } 小时是否需要填充0, 时间
 */
export function getGMTTimezone (timezone: string, {
  padNumber = false,
  date = moment(),
}: {
  padNumber?: boolean;
  date?: moment.Moment;
} = {
  padNumber: false,
  date: moment(),
}): string {
  const fullOffset = moment(date).tz(timezone).format('Z')
  if (padNumber) {
    return `GMT${fullOffset}`
  }
  return `GMT${fullOffset.replace(':00', '').replace('+0', '+').replace('-0', '-')}`
}

/**
 * @description 根据moment对象时间获得Julian day
 * @param momentDate 
 */
export function julianDay (momentDate: moment.Moment): number {
  let y = momentDate.year()
  let m = momentDate.month() + 1
  const d = momentDate.date()
  const a = Math.floor((14 - m) / 12)
  y = y + 4800 - a
  m = m + 12 * a - 3
  return (
    d +
    Math.floor((153 * m + 2) / 5) +
    365 * y +
    Math.floor(y / 4) -
    Math.floor(y / 100) +
    Math.floor(y / 400) -
    32045
  )
}
