// @lark/moment-timezone/src/standards.ts
/**
 * https://bytedance.feishu.cn/sheets/shtcn6NEOBFdwpCLxYTzVpFwsR0#2ef303
 */
import moment from '@lark/moment-timezone-polyfill'
import { getLocalTimezone, supportLocales } from './core'
import { getLocalNow, getMomentByUnix } from './extend'
import { formatMoment, getGMTTimezone, __T, IFormatOption } from './format'
import platform from './platform'

type TimePreciseness = 'hour' | 'minute' | 'second'
type TimePrecisenessOptions = {
  preciseness?: TimePreciseness;
  isZeroMinute?: boolean;
}
type DatePreciseness = 'month' | 'day'
type DatePrecisenessOptions = {
  preciseness?: DatePreciseness;
}

type FormatLengthType = 'long' | 'short'

type Unix = number
type TimeCompatible = moment.Moment | Date | Unix

const defaultLocale = 'en-US'
const localeMap = supportLocales.reduce((acc, key) => ({
  ...acc,
  [key.substr(0, 2)]: key,
}), {})

/**
 * i18n
 */
function i18n (key: string, locale?: string) {
  const localeCode = (locale || moment.locale() || defaultLocale).replace(/(\w+)[-_](\w+)/, function (m, $1, $2) {
    return `${$1}-${$2.toUpperCase()}`
  })
  const i18ns = platform[localeCode] || platform[localeMap[localeCode]]
  if (!i18ns) return undefined
  return i18ns[key]
}
function i18nTpl (key: string, data?: {}, { locale, is12Hour }: Pick<IFormatOption, 'locale' | 'is12Hour'> = {}) {
  const localeString = i18n(key, locale)
  if (!localeString) return key
  return __T(localeString, data, is12Hour)
}

function defineFormatterByI18nKey (key: string) {
  return function (date: moment.Moment, option: IFormatOption = {}, context = {}) {
    return formatMoment(date, i18n(key, option.locale), context, option)
  }
}

function defineCombinerByI18nKey<T extends {}> (key: string) {
  return function (data: T, option: IFormatOption = {}) {
    return i18nTpl(key, data, option)
  }
}

function getMoment (date: TimeCompatible, { locale, timezone }: IFormatOption = {}): moment.Moment | null {
  if (!date) return null
  let targetMoment: moment.Moment
  if (date instanceof Date) {
    targetMoment = moment(date)
  } else if (typeof date === 'number') {
    targetMoment = moment(date * 1000)
  } else {
    targetMoment = date.clone()
  }
  if (timezone) targetMoment = targetMoment.tz(timezone)
  if (locale) targetMoment = targetMoment.locale(locale)
  return targetMoment
}

function compatibleMoment<R extends any, O extends IFormatOption> (fn: (date: moment.Moment, options?: O) => R) {
  return function (date: TimeCompatible, options?: O): R {
    if (typeof options === 'object' && options.timezone) {
      options = Object.assign({}, options, {
        displayTimezone: options.timezone,
      })
    }
    if (typeof options === 'object' && options.locale && !isSupportedLocale(options.locale)) {
      options = Object.assign({}, options, {
        locale: moment.locale() || defaultLocale,
      })
    }
    const time = getMoment(date, options)
    if (!time) return '' as any as R
    return fn(time, options)
  }
}

function compatibleMomentRange<R, O extends IFormatOption> (fn: (start: moment.Moment, end: moment.Moment, options?: O) => R) {
  return function (start: TimeCompatible, end: TimeCompatible, options?: O): R {
    if (typeof options === 'object' && options.timezone) {
      options = Object.assign({}, options, {
        displayTimezone: options.timezone,
      })
    }
    const startTime = getMoment(start, options)
    const endTime = getMoment(end, options)
    if (!startTime || !endTime) return '' as any as R
    return fn(startTime, endTime, options)
  }
}

/**
 * ========================= Util ===================
 */

export function isSupportedLocale (locale: string) {
  return Boolean(platform[locale] || platform[localeMap[locale]])
}


/**
 * ========================= Time Format ===================
 */

/**
 * 07:30:28 |	7:30:28 AM
 */
export const formatLongSecondTime = compatibleMoment(defineFormatterByI18nKey('LongSecondFormat'))

/**
 * 07:00 | 7:00 AM
 */
export const formatShortMinuteTime = compatibleMoment(defineFormatterByI18nKey('ShortMinuteFormat'))

/**
 * 没有上午/下午的时间:
 * 07:00 | 7:00
 */
export const formatMinMinuteTime = compatibleMoment(defineFormatterByI18nKey('MinMinuteFormat'))

/**
 * en: 07:20 |	7 AM
 * zh: 07:20 |  上午7点
 */
export const formatShortHourTime = compatibleMoment(defineFormatterByI18nKey('ShortHourFormat'))

/**
 * 没有上下午
 * en: 13:20 |	1
 * zh: 13:20 |	1点
 */
export const formatMinHourTime = compatibleMoment(defineFormatterByI18nKey('MinHourFormat'))

/**
 * minute固定为00
 * en: 07:00 |	7 AM
 * zh: 07:00 |  上午7点
 */
export const formatZeroMinuteHourTime = compatibleMoment(
  defineFormatterByI18nKey('ZeroMinuteHourFormat')
)
/**
 * ========================= Meridiem Format ===================
 */

/**
 * AM/PM
 */
export const formatMeridiem = compatibleMoment(defineFormatterByI18nKey('MeridiemFormat'))

/**
 * ========================= Date Format ===================
 */

/**
 * en: Mar 15
 * zh: 3月15日
 */
export const formatShortDate = compatibleMoment(defineFormatterByI18nKey('ShortDateFormat'))

/**
 * en: Mar 15, 2019
 * zh: 2019年3月15日
 */
export const formatLongDate = compatibleMoment(defineFormatterByI18nKey('LongDateFormat'))

/**
 * ========================= Weekday Format ===================
 */

/**
 * en: Thu
 */
export const formatShortWeekday = compatibleMoment(defineFormatterByI18nKey('ShortWeekdayFormat'))

/**
 * en: Thursday
 */
export const formatLongWeekday = compatibleMoment(defineFormatterByI18nKey('LongWeekdayFormat'))


/**
 * ========================= Month Format ===================
 */

/**
* en: March
* zh: 3月
*/
export const formatShortMonth = compatibleMoment(defineFormatterByI18nKey('ShortMonthFormat'))

/**
* en: Mar 2019
* zh: 2019年3月
*/
export const formatLongMonth = compatibleMoment(defineFormatterByI18nKey('LongMonthFormat'))

/**
 * ========================= Combine Format ===================
 */

/**
  * {{date}} {{time}}
  */
export const combineDateTime = defineCombinerByI18nKey<{
  date: string;
  time: string;
}>('DateTimeCombineFormat')

/**
*  day: week day / relative day
* {{day}}, {{date}} | {{date}} ({{day}})
*/
export const combineDateDay = defineCombinerByI18nKey<{
  date: string;
  day: string;
}>('DateDayCombineFormat')

/**
* {{time}} ({{GMT}})
*/
export const combineGMT = defineCombinerByI18nKey<{
  time: string;
  GMT: string;
}>('GMTTimeFormat')

/**
 * ====================== Time Range ========================
 */

/**
 * {{startTime}} - {{endTime}}
 */
export const combineTimeRange = defineCombinerByI18nKey<{
  startTime: string;
  endTime: string;
}>('TimeRangeFormat')

/**
 *  处理共用上/下午的语序，如英文
 * {{minStartTime}} - {{minEndTime}} {{meridiem}}
 */
export const combineTimeRangeInSameMeridiem = defineCombinerByI18nKey<{
  minStartTime: string;
  minEndTime: string;
  meridiem: string;
}>('TimeRangeInSameMeridiemFormat')

/**
 *  处理共用某一个日期
 * {{date}} {{rangeTime}}
 */
export const combineTimeRangeInOneDay = defineCombinerByI18nKey<{
  date: string;
  rangeTime: string;
}>('TimeRangeInOneDayFormat')

/**
 * ========================= smart standard formats based on current time ======================
 */

export const isToday = compatibleMoment(function (date: moment.Moment) {
  return date.isSame(getLocalNow(), 'day')
})

/**
 * 是否在昨天/今天/明天
 */
export const isInRecentThreeDays = compatibleMoment(function (date: moment.Moment) {
  const timezone = date.tz()
  const nowMoment = timezone ? moment().tz(timezone) : getLocalNow()
  const yesterday = nowMoment.clone().subtract(1, 'days').startOf('day')
  const tomorrow = nowMoment.clone().add(1, 'days').endOf('day')
  // 昨天/今天/明天
  return !date.isBefore(yesterday) && !date.isAfter(tomorrow)
})
/**
 * 最近三天：昨天/今天/明天
 * 其他： 周几
 */
export const fortmatRelativeDay = compatibleMoment(function (date: moment.Moment) {
  return isInRecentThreeDays(date) ? date.calendar() : formatShortWeekday(date)
})

/**
 *          12h                               24h
 * zh    下午5点  下午5:08  下午5:08:00 |  17:00   17:08   17:08:00
 * en    5 PM    5:08 PM  5:08:00 PM |  17:00   17:08   17:08:00
 *
 */
export const formatTime = compatibleMoment(function (date: moment.Moment, {
  preciseness = 'minute',
  abbrIntHour = false,
  isZeroMinute = false,
  ...formatOption
}: IFormatOption & TimePrecisenessOptions & {
  abbrIntHour?: boolean;
} = {}) {
  if (!date) return ''
  const isInt = date.minutes() === 0 && date.seconds() === 0
  const formatPreciseness = abbrIntHour && isInt ? 'hour' : preciseness
  if (formatPreciseness === 'second') {
    return formatLongSecondTime(date, formatOption)
  }
  if (formatPreciseness === 'hour') {
    return isZeroMinute
      ? formatZeroMinuteHourTime(date, formatOption)
      : formatShortHourTime(date, formatOption)
  }
  return formatShortMinuteTime(date, formatOption)
})

/**
 * ----- day --------
 * 在今年： 7月9日
 * 否则： 2019年7月9日
 * ----- month ------
 * 在今年： 7月
 * 否则：2019年7月
 * -------------------
 * short比long少了年份
 */
export const formatDate = compatibleMoment(function (date: moment.Moment, {
  abbrCurrentYear = true,
  preciseness = 'day',
  type = 'long',
  ...formatOption
}: IFormatOption & DatePrecisenessOptions & {
  abbrCurrentYear?: boolean;
  type?: FormatLengthType;
} = {}) {
  const nowMoment = getLocalNow()
  const useShortFormat = type === 'short' ||
    abbrCurrentYear && nowMoment.isSame(date, 'year')
  if (useShortFormat) {
    return preciseness === 'month' ? formatShortMonth(date, formatOption) : formatShortDate(date, formatOption)
  }
  return preciseness === 'month' ? formatLongMonth(date, formatOption) : formatLongDate(date, formatOption)
})

/**
 * 6月9日（周二）
 * 2019年6月9日（周二）
 * 与formatRelativeFullDate差异在于这是个静态的，不会存在“今天”/“明天”这种
 */
export const formatFullDate = compatibleMoment(function (date: moment.Moment, {
  abbrCurrentYear = true,
  type = 'long',
  ...formatOption
}: IFormatOption & {
  abbrCurrentYear?: boolean;
  type?: FormatLengthType;
} = {}) {
  return combineDateDay({
    date: formatDate(date, { abbrCurrentYear, type, ...formatOption }),
    day: formatShortWeekday(date, formatOption),
  }, formatOption)
})

/**
 *     24h                        12h
 * 3月15日 13:05	      |       3月15日 下午1:05
 * 2019年3月15日 13:05	|      2019年3月15日 下午1:05
 */
export const formatDateTime = compatibleMoment(function (date: moment.Moment, {
  preciseness = 'minute',
  abbrCurrentYear = true,
  type = 'long',
  ...formatOption
}: IFormatOption & TimePrecisenessOptions & {
  abbrCurrentYear?: boolean;
  type?: FormatLengthType;
} = {}) {
  return combineDateTime({
    date: formatDate(date, { abbrCurrentYear, type, ...formatOption }),
    time: formatTime(date, { preciseness, ...formatOption }),
  }, formatOption)
})

/**
 * abbrCurrentYear: true
 * 6月9日（周二）14:00（GMT+8）
 * 2019年6月9日（周二）下午2:00 （GMT+8）
 */
export const formatFullDateTime = compatibleMoment(function (
  date: moment.Moment,
  {
    abbrCurrentYear = false,
    preciseness = 'minute',
    displayTimezone = getLocalTimezone(),
    showGMT = true,
    type = 'long',
    ...formatOption
  }: IFormatOption & TimePrecisenessOptions & {
    abbrCurrentYear?: boolean;
    displayTimezone?: string;
    showGMT?: boolean;
    type?: FormatLengthType;
  } = {}
) {
  const dateString = formatFullDate(date, { abbrCurrentYear, type, ...formatOption })
  const timeString = formatTime(date, { preciseness, ...formatOption })
  const dateTimeString = combineDateTime({
    date: dateString,
    time: timeString,
  }, formatOption)
  if (!showGMT) return dateTimeString
  return combineGMT({
    time: dateTimeString,
    GMT: getGMTTimezone(displayTimezone, { date }),
  }, formatOption)
})

/**
 * ==================== foramt Relative Time ======================
 */

/**
 * 昨天/今天/明天
 * 3月15日
 * 2019年3月15日
 */
export const formatRelativeDate = compatibleMoment(function (date: moment.Moment, formatOption: IFormatOption = {}) {
  return isInRecentThreeDays(date, formatOption) ?
    fortmatRelativeDay(date, formatOption) : formatDate(date, formatOption)
})

/**
 * 6月9日（周二）
 * 6月9日（昨天）
 * 2019年6月9日（周二）
 */
export const formatRelativeFullDate = compatibleMoment(function (date: moment.Moment, {
  abbrCurrentYear = true,
  type = 'long',
  ...formatOption
}: IFormatOption & {
  abbrCurrentYear?: boolean;
  type?: FormatLengthType;
} = {}) {
  return combineDateDay({
    date: formatDate(date, { abbrCurrentYear, type, ...formatOption }),
    day: fortmatRelativeDay(date, formatOption),
  }, formatOption)
})

/**
 * 6月9日（周二）14:00（GMT+8）
 * 6月9日（今天）14:00（GMT+8）
 * 2019年6月9日（周二）下午2:00 （GMT+8）
 */
export const formatRelativeFullDateTime = compatibleMoment(function (
  date: moment.Moment,
  {
    abbrCurrentYear = true,
    preciseness = 'minute',
    displayTimezone = getLocalTimezone(),
    showGMT = true,
    type = 'long',
    ...formatOption
  }: IFormatOption & TimePrecisenessOptions & {
    abbrCurrentYear?: boolean;
    displayTimezone?: string;
    showGMT?: boolean;
    type?: FormatLengthType;
  } = {}
) {
  const formatDate = date.tz(displayTimezone)
  const dateString = formatRelativeFullDate(formatDate, { abbrCurrentYear, type, ...formatOption })
  const timeString = formatTime(formatDate, { preciseness, ...formatOption })
  const dateTimeString = combineDateTime({
    date: dateString,
    time: timeString,
  }, formatOption)
  if (!showGMT) return dateTimeString
  return combineGMT({
    time: dateTimeString,
    GMT: getGMTTimezone(displayTimezone, { date }),
  }, formatOption)
})

/**
 *
 * ------------------ short --------------------------------
 *  24h                        12h
 * <1d: 13:05             <1d: 下午1:05
 * <2d: 昨天               <2d: 昨天
 * <1year: 3月15日         <1year: 3月15日
 * Else: 2019年3月15日     Else: 2019年3月15日
 * ----------------- long -----------------------------------
 *  24h                               12h
 * <1d: 13:05                    <1d: 下午1:05
 * <2d: 昨天 13:05                <2d: 昨天 下午1:05
 * <1year: 3月15日 13:05          <1year: 3月15日 下午1:05
 * Else: 2019年3月15日 13:05      Else: 2019年3月15日 下午1:05
 * https://bytedance.feishu.cn/docs/doccn29NCH4KJYO2lVMCqe5a7Jd#
 */
export const formatRelativeDateTime = compatibleMoment(function (date: moment.Moment, {
  preciseness = 'minute',
  type = 'long',
  ...formatOption
}: IFormatOption & TimePrecisenessOptions & {
  type?: 'short' | 'long';
} = {}) {
  if (isToday(date)) {
    return formatTime(date, { preciseness, ...formatOption })
  }

  if (type === 'short') {
    return formatRelativeDate(date, formatOption)
  }
  return combineDateTime({ date: isInRecentThreeDays(date, formatOption) ? fortmatRelativeDay(date, formatOption) : formatDate(date, formatOption),
    time: formatTime(date, { preciseness, ...formatOption }),
  }, formatOption)
})

/**
 * ==================== foramt Time Range ======================
 */

/**
 * trimZeroMinutes=true: 上午5点 - 9点    5 - 9 AM
 * trimZeroMinutes=false: 上午5:00 - 9: 00   5:00 - 9:00 AM
 * trimZeroMinutes=true: 上午5点 - 9点    5 - 9 AM
 * trimZeroMinutes=false: 上午5:00 - 下午1:00  5:00 AM - 1:00 PM
 */
export const formatTimeRange = compatibleMomentRange(function (
  startTime: moment.Moment,
  endTime: moment.Moment,
  {
    abbrIntHour = false,
    preciseness = 'minute',
    displayTimezone = getLocalTimezone(),
    showGMT = false,
    ...formatOption
  }: IFormatOption & TimePrecisenessOptions & {
    abbrIntHour?: boolean;
    displayTimezone?: string;
    showGMT?: boolean;
  } = {}) {
  const startGMTTimezone = getGMTTimezone(displayTimezone, { date: startTime })
  const endGMTTimezone = getGMTTimezone(displayTimezone, { date: endTime })
  const showStartGMT = showGMT && startGMTTimezone !== endGMTTimezone

  const startTimeMeridiem = formatMeridiem(startTime, formatOption)
  const endTimeMeridiem = formatMeridiem(endTime, formatOption)

  let start = formatTime(startTime, { abbrIntHour, preciseness, ...formatOption })
  let end = formatTime(endTime, { abbrIntHour, preciseness, ...formatOption })
  if (showStartGMT) {
    start = combineGMT({ time: start, GMT: startGMTTimezone })
  }
 
  let time
  if (startTimeMeridiem === endTimeMeridiem && !showStartGMT) {
    start = start.replace(startTimeMeridiem, '').trim()
    end = end.replace(endTimeMeridiem, '').trim()
    time = combineTimeRangeInSameMeridiem({
      minStartTime: start,
      minEndTime: end,
      meridiem: startTimeMeridiem,
    }, formatOption)
  } else {
    time = combineTimeRange({
      startTime: start,
      endTime: end,
    })
  }
  if (showGMT) {
    time = combineGMT({ time, GMT: endGMTTimezone })
  }
  return time
})

/**
 * 2019年8月1日
 */
export const formatDateRange = compatibleMomentRange(function (
  startDate: moment.Moment,
  endDate: moment.Moment,
  {
    displayTimezone = getLocalTimezone(),
    showGMT = false,
    ...formatOption
  }: IFormatOption & {
    displayTimezone?: string;
    showGMT?: boolean;
  } = {}) {
  const now = getLocalNow()
  const showShort = startDate.isSame(now, 'year') &&
    endDate.isSame(now, 'year')
  const format = date => combineDateDay({
    date: formatDate(date, { abbrCurrentYear: showShort, ...formatOption }),
    day: fortmatRelativeDay(date, formatOption),
  }, formatOption)

  const dateRangeString = combineTimeRange({
    startTime: format(startDate),
    endTime: format(endDate),
  }, formatOption)
  return showGMT ? combineGMT({
    time: dateRangeString,
    GMT: getGMTTimezone(displayTimezone, { date: startDate }),
  }, formatOption) : dateRangeString
})

/**
 * 2019年8月1日（周二）下午3点 - 5点
 * 2019年8月1日（周二）下午3:05 - 5:08
 * 8月1日（周二）下午3:05 - 8月3日（周三）上午5:08
 */
export const formatDateTimeRange = compatibleMomentRange(function (
  startTime: moment.Moment,
  endTime: moment.Moment,
  {
    preciseness = 'minute',
    displayTimezone = getLocalTimezone(),
    showGMT = false,
    ...formatOption
  }: IFormatOption & {
    displayTimezone?: string;
    showGMT?: boolean;
    preciseness?: TimePreciseness;
  } = {}) {
  if (startTime.isSame(endTime, 'day')) {
    return combineDateTime({
      date: formatRelativeFullDate(startTime, formatOption),
      time: formatTimeRange(startTime, endTime, { ...formatOption, displayTimezone, showGMT }),
    }, formatOption)
  }

  const now = getLocalNow()
  const abbrCurrentYear = now.isSame(startTime, 'year') && now.isSame(endTime, 'year')

  const startGMTTimezone = getGMTTimezone(displayTimezone, { date: startTime })
  const endGMTTimezone = getGMTTimezone(displayTimezone, { date: endTime })
  const showStartGMT = showGMT && startGMTTimezone !== endGMTTimezone

  return combineTimeRange({
    startTime: formatRelativeFullDateTime(startTime, {
      abbrCurrentYear, preciseness, showGMT: showStartGMT, displayTimezone,
      ...formatOption,
    }),
    endTime: formatRelativeFullDateTime(endTime, {
      abbrCurrentYear, preciseness, showGMT, displayTimezone,
      ...formatOption,
    }),
  }, formatOption)
})

/**
 * ======================== Calendar Biz ==================
 */

/**
 * -----全天------
 * 6月9日（周二）
 * 6月9日（昨天）
 * 2019年6月9日（周二）
 * -----非全天------
 * 6月9日（周二）14:00（GMT+8）
 * 2019年6月9日（周二）下午2:00 （GMT+8）
 */
export const formatEventTimeString = compatibleMoment(function (
  date: moment.Moment,
  {
    displayTimezone = getLocalTimezone(),
    isAllDay = false,
    ...formatOption
  }: IFormatOption & {
    displayTimezone?: string;
    isAllDay?: boolean;
  } = {}
) {
  return isAllDay ? formatRelativeFullDate(date, formatOption) : formatRelativeFullDateTime(date, {
    displayTimezone,
    showGMT: true,
    ...formatOption,
  })
})

/**
 * 获取提醒的时间，跟日程的时间差别在“今天”没有日期
 * Today 7:22 PM(GMT+8)
 */
export const formatEventReminderTimeString = compatibleMoment(function (
  date: moment.Moment,
  {
    displayTimezone = getLocalTimezone(),
    isAllDay = false,
    ...formatOption
  }: IFormatOption & {
    displayTimezone?: string;
    isAllDay?: boolean;
  } = {}
) {
  if (date.isSame(getLocalNow(), 'day')) {
    return combineGMT({
      time: formatRelativeDateTime(date),
      GMT: getGMTTimezone(displayTimezone, { date }),
    }, formatOption)
  }
  return formatEventTimeString(date, {
    isAllDay, displayTimezone,
    ...formatOption,
  })
})

/**
 * 获取开始结束时间range描述字符串
 */
export function formatEventTimeRangeString ({
  startTime,
  startTimezone,
  endTime,
  endTimezone,
  isAllDay = false,
  displayTimezone = getLocalTimezone(),
  showGMT = true,
}: {
  startTime: number;
  startTimezone: string;
  endTime: number;
  endTimezone: string;
  isAllDay?: boolean;
  displayTimezone?: string;
  showGMT?: boolean;
}, formatOption: IFormatOption = {}) {
  // 全天日程
  if (isAllDay) {
    const startTimeMoment = getMomentByUnix({
      timestamp: startTime,
      timezone: startTimezone,
      isAllDay: true,
    }).tz(displayTimezone)
    const endTimeMoment = getMomentByUnix({
      timestamp: endTime,
      timezone: endTimezone,
      isAllDay: true,
    }).add(-1, 'seconds')
      .tz(displayTimezone)
    if (endTimeMoment.isSame(startTimeMoment, 'day')) {
      return formatRelativeFullDate(startTimeMoment, formatOption)
    }
    return formatDateRange(startTimeMoment, endTimeMoment, formatOption)
  }
  // 非全天日程
  const startTimeMoment = getMomentByUnix({
    timestamp: startTime,
    timezone: displayTimezone,
    isAllDay: false,
  })
  const endTimeMoment = getMomentByUnix({
    timestamp: endTime,
    timezone: displayTimezone,
    isAllDay: false,
  })
  return formatDateTimeRange(startTimeMoment, endTimeMoment, {
    displayTimezone,
    showGMT,
    ...formatOption,
  })
}

/**
 * 同一天的两个时间，
 * if: 上午8点 - 9点  return {startLabel: "上午8点", endLabel: "9点" }
 * output: 8 - 9 AM  return {startLabel: "8", endLabel: "9 AM"}
 */
export function formatEventRangeTimeLablels ({
  startTime,
  endTime,
  abbrIntHour = false,
}: {
  startTime: moment.Moment;
  endTime: moment.Moment;
  abbrIntHour: boolean;
}, formatOption: IFormatOption = {}) {
  let trimStartMeridiem = false
  let trimEndMeridiem = false
  if (startTime && endTime && startTime.format('A') === endTime.format('A')) {
    // 分析meridiem和time，英文模式显示   8 - 9 AM
    const formatString = i18n('TimeRangeInSameMeridiemFormat')
    const AIdx = formatString.indexOf('{{meridiem}}')
    const hIdx = formatString.indexOf('{{minStartTime}}')
    if (AIdx > hIdx) {
      trimStartMeridiem = true
    } else {
      trimEndMeridiem = true
    }
  }
  const startLabel = startTime ? formatTime(startTime, {
    abbrIntHour,
    preciseness: 'minute',
    ...formatOption,
  }) : ''
  const endLabel = endTime ? formatTime(endTime, {
    abbrIntHour,
    preciseness: 'minute',
    ...formatOption,
  }) : ''
  return {
    startLabel: trimStartMeridiem ? startLabel.replace(formatMeridiem(startTime, formatOption), '').trim() : startLabel,
    endLabel: trimEndMeridiem ? endLabel.replace(formatMeridiem(endTime, formatOption), '').trim() : endLabel,
  }
}
