// @universe-design/react/src/badge/stories/homeless.stories.tsx
import type { Meta } from '@storybook/react';
import React, { forwardRef } from 'react';
import { Badge } from '../..';
import { CanvasTextBadgeSidecar } from '../CanvasTextBadge';
import { PerformanceContent } from '../../_stories/common/PerformanceContent';
import { SVGTextBadgeSidecar } from '../SVGTextBadge';
import { LogsContainer } from '../../_stories/common/LogsContainer';

export default {
  title: 'Components/Badge/临时示例',
  component: Badge,
} satisfies Meta;

const Box = forwardRef<HTMLDivElement>((_, ref) => (
  <div
    ref={ref}
    style={{
      width: 42,
      height: 42,
      backgroundColor: '#EEEEEE',
      borderRadius: '50%',
    }}
  />
));

const count = 2000;
const randomContent = true;
export const Performance = () => (
  <>
    <LogsContainer />
    <PerformanceContent
      componentName="DOM 普通布局模式"
      count={count}
      itemRenderer={i => (
        <Badge
          size="large"
          {...{
            count: randomContent ? String(Math.round(Math.random() * count)) : String(count),
          }}
        >
          <Box />
        </Badge>
      )}
    />
    <PerformanceContent
      componentName="DOM 高性能布局模式"
      count={count}
      itemRenderer={i => (
        <Badge
          childCornerRadius={21}
          size="large"
          {...{
            count: randomContent ? String(Math.round(Math.random() * count)) : String(count),
          }}
        >
          <Box />
        </Badge>
      )}
    />
    <PerformanceContent
      componentName="Canvas 高性能布局模式"
      count={count}
      itemRenderer={i => (
        <Badge
          childCornerRadius={21}
          textBadgeSidecar={CanvasTextBadgeSidecar}
          size="large"
          {...{
            count: randomContent ? String(Math.round(Math.random() * count)) : String(count),
          }}
        >
          <Box />
        </Badge>
      )}
    />
    <PerformanceContent
      componentName="SVG 高性能布局模式"
      count={count}
      itemRenderer={i => (
        <Badge
          childCornerRadius={21}
          textBadgeSidecar={SVGTextBadgeSidecar}
          size="large"
          {...{
            count: randomContent ? String(Math.round(Math.random() * count)) : String(count),
          }}
        >
          <Box />
        </Badge>
      )}
    />
    <PerformanceContent
      componentName="DOM 渲染 SVG"
      count={count}
      itemRenderer={i => (
        <Badge
          childCornerRadius={21}
          size="large"
          {...{
            count: (
              <svg width={30} height={14} viewBox="0 0 14 30">
                <text textAnchor="center" x="50%" y="50%" dominantBaseline="middle" fill="white">
                  {randomContent ? String(Math.round(Math.random() * count)) : String(count)}
                </text>
              </svg>
            ),
          }}
        >
          <Box />
        </Badge>
      )}
    />
  </>
);
