// @universe-design/react/src/badge/stories/screenshot.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C999Cases_0 from '../demo/999.cases-0';
import C999Cases_1 from '../demo/999.cases-1';
import CIndependentUseTextBadge from '../demo/independent-use-text-badge';
import CZCanvasTextBadgeScreenshot from '../demo/z.canvas-text-badge.screenshot';
if (!(C999Cases_0 as Story).storyName) {
  (C999Cases_0 as Story).storyName = '999 cases 0';
}
if (!(C999Cases_1 as Story).storyName) {
  (C999Cases_1 as Story).storyName = '999 cases 1';
}
if (!(CIndependentUseTextBadge as Story).storyName) {
  (CIndependentUseTextBadge as Story).storyName = 'Independent use text badge';
}
if (!(CZCanvasTextBadgeScreenshot as Story).storyName) {
  (CZCanvasTextBadgeScreenshot as Story).storyName = 'Z canvas text badge screenshot';
}

export default {
  title: 'Components/Badge/截图测试',
  tags: ['autodocs'],
};

export { C999Cases_0, C999Cases_1, CIndependentUseTextBadge, CZCanvasTextBadgeScreenshot };
