// @universe-design/react/src/badge/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Basic from '../demo/0.basic';
import C1_1Dot from '../demo/1.1.dot';
import C1_2Size from '../demo/1.2.size';
import C1Position from '../demo/1.position';
import C2OverflowCount from '../demo/2.overflow-count';
import C3Status from '../demo/3.status';
import C4Colorful from '../demo/4.colorful';
import C5Hack10px from '../demo/5.hack10px';
import CCollaboration from '../demo/collaboration';
if (!(C0Basic as Story).storyName) {
  (C0Basic as Story).storyName = '0. 基础用法';
  (C0Basic as Story).parameters = {
    docs: {
      description: {
        story: '简单的用法。当`count`为 0 时，默认不显示，但是可以使用`showZero`设置为显示。\n\n',
      },
    },
  };
}
if (!(C1_1Dot as Story).storyName) {
  (C1_1Dot as Story).storyName = '1.1. 点状徽标';
  (C1_1Dot as Story).parameters = {
    docs: {
      description: {
        story:
          '没有具体的数字，只表示未完成状态。内容为文字时，徽标左边与文字右边距离 2px，徽标顶端与文字顶端平齐，可以将 `contentType` 属性设置为 `text`。\n\n',
      },
    },
  };
}
if (!(C1_2Size as Story).storyName) {
  (C1_2Size as Story).storyName = '1.2. 大小';
  (C1_2Size as Story).parameters = {
    docs: {
      description: {
        story: '点状徽标可以通过`literal`类型的`size`指定徽标大小，也可以通过`number`类型的`size`指定徽标大小。\n\n',
      },
    },
  };
}
if (!(C1Position as Story).storyName) {
  (C1Position as Story).storyName = '1. 位置';
  (C1Position as Story).parameters = {
    docs: {
      description: {
        story: '徽标具有内容元素时，一共有四种方位布局。\n\n',
      },
    },
  };
}
if (!(C2OverflowCount as Story).storyName) {
  (C2OverflowCount as Story).storyName = '2. 封顶数字';
  (C2OverflowCount as Story).parameters = {
    docs: {
      description: {
        story:
          '超过`overflowCount`值会显示为`...`，默认的`overflowCount`为 999。如果需要自定义文案，则直接传入非`number`类型的`content`即可。\n\n',
      },
    },
  };
}
if (!(C3Status as Story).storyName) {
  (C3Status as Story).storyName = '3. 状态徽标';
  (C3Status as Story).parameters = {
    docs: {
      description: {
        story: '用于表示状态的小圆点。\n\n',
      },
    },
  };
}
if (!(C4Colorful as Story).storyName) {
  (C4Colorful as Story).storyName = '4. 炫彩徽标';
  (C4Colorful as Story).parameters = {
    docs: {
      description: {
        story:
          '我们添加了多种预设色彩的徽标样式，用作不同场景使用。如果预设值不能满足你的需求，可以设置为具体的色值。\n\n',
      },
    },
  };
}
if (!(C5Hack10px as Story).storyName) {
  (C5Hack10px as Story).storyName = '5. 在默认最小字号小于 10px 的浏览器中运行';
  (C5Hack10px as Story).parameters = {
    docs: {
      description: {
        story:
          '目前主流浏览器中 **Safari** **Firefox** 不限制最小字号，**new Edge** **Chrome** **Vulcan** **Chromium** 使用 **12px** 作为最小字号。我们提供一种兼容模式使用 `font-size: 12px; zoom: 0.8333` 模拟 **10px** 字号，可以使用 `hack10px` 属性启用。\n\n',
      },
    },
  };
}
if (!(CCollaboration as Story).storyName) {
  (CCollaboration as Story).storyName = '5. 带描边的协同状态徽标';
  (CCollaboration as Story).parameters = {
    docs: {
      description: {
        story: '<code src="./collaboration.tsx" />\n',
      },
    },
  };
}

export default {
  title: 'Components/Badge/官网示例',
  tags: ['autodocs'],
};

export { C0Basic, C1_1Dot, C1_2Size, C1Position, C2OverflowCount, C3Status, C4Colorful, C5Hack10px, CCollaboration };
