// @universe-design/react/src/transfer/hooks/PanelContext.ts
import type * as React from 'react';
import { createContextHook } from '@universe-design/react-utilities';
import type { TransferOption, TransferValue } from '../interface';

export type TransferPanelContextType<
  V extends TransferValue = string,
  OP extends TransferOption<V> = TransferOption<V>,
> = {
  /**
   * 空状态的内容
   */
  emptyContent?: React.ReactNode;
  /**
   * 自定义选项渲染
   */
  optionRenderer?: (params: {
    /**
     * 选项详情
     */
    option: OP;
    /**
     * 是否已经被选中
     */
    isChecked: boolean;
    /**
     * toggle
     */
    toggle: (value: V) => void;
  }) => React.ReactElement;
};

const { useContext: _useTransferPanelContext, Provider: TransferPanelProvider } =
  createContextHook<TransferPanelContextType>('TransferContext', {
    strict: true,
  });

export const useTransferPanelContext = _useTransferPanelContext as <
  V extends TransferValue = string,
  OP extends TransferOption<V> = TransferOption<V>,
>() => TransferPanelContextType<V, OP>;
export { TransferPanelProvider };
