// @universe-design/react/src/transfer/hooks/TransferContext.tsx
import React from 'react';
import { useCssScope } from '../../config-provider';
import { createContextHook, useControlledState, useConstant } from '@universe-design/react-utilities';
import { useTransferData } from './useTransferData';
import type { TransferGroupOption, TransferOption, TransferValue, UseTransferProps } from '../interface';
import { TransferChangeAction } from '../interface';
import { defaultFilterOption } from '../utils';

/**
 * @public
 */
export interface TransferContextType<
  V extends TransferValue = string,
  OP extends TransferOption<V> = TransferOption<V>,
> {
  /**
   * cssScope
   */
  cssScope: (suffix: string) => string;
  /**
   * @internal
   * 切换全选状态
   */
  toggleSelectAll: () => void;
  /**
   * 已选值
   */
  value: V[];
  /**
   * 设置已选值
   */
  setValue: (newValue: V[], action: TransferChangeAction) => void;
  /**
   * 已选值集合
   * @internal
   */
  valueSet: Set<V>;
  /**
   * 已选值对应的 options
   */
  selectedOptions: OP[];
  /**
   * 清除全部
   */
  clearAll: () => void;
  /**
   * 根据 searchValue 筛选后的值
   */
  filteredOptions: OP[];
  /**
   * 输入的搜索值
   */
  searchValue: string;
  /**
   * 切换某一选项
   */
  toggle: (v: V) => void;
  /**
   * 设置搜索条件
   */
  setSearchValue: (searchValue: string) => void;
  /**
   * 当前筛选条件下的选项是否已经全部选中
   */
  isAllSelected: boolean;
  /**
   * 所有分组信息（key: 分组的 groupKey, value: 分组 ）
   */
  groupWeakMap: WeakMap<OP, TransferGroupOption<V, OP>>;
  /**
   * 是否显示分组信息
   */
  group: boolean | [boolean, boolean];
}

export function useTransfer<V extends TransferValue = string, OP extends TransferOption<V> = TransferOption<V>>(
  props: UseTransferProps<V, OP>,
): TransferContextType<V, OP> {
  const {
    value: valueProp,
    defaultValue,
    options = [],
    onChange,
    onClear,
    onSearch,
    onSelect,
    onDeSelect,
    onToggleSelectAll,
    filterOption = defaultFilterOption,
    group = false,
    defaultSearchValue,
  } = props;

  const cssScope = useCssScope('transfer');

  const [searchValue, _setSearchValue] = React.useState(defaultSearchValue || '');

  const [value, _setValue] = useControlledState<V[]>(defaultValue ?? [], valueProp, {
    isControlledOverride: 'value' in props,
  });
  const selectedCache = useConstant<Map<V, OP>>(() => new Map());

  const { optionsMap, filteredOptions, groupWeakMap } = useTransferData<V, OP>({
    options,
    value,
    searchValue,
    filterOption,
    group,
  });

  const setValue = React.useCallback(
    (newValue: V[], action: TransferChangeAction) => {
      newValue.forEach(v => {
        const option = optionsMap.get(v);
        if (!option) {
          return;
        }
        // 实时更新 optionsMap 防止下次 render 时 options 被变更
        selectedCache.set(v, option);
      });
      const _options = newValue.map(v => selectedCache.get(v));
      onChange?.(newValue, { action, options: _options as OP[] });
      _setValue(newValue);
    },
    [_setValue, onChange, optionsMap, selectedCache],
  );

  const valueSet = React.useMemo(() => new Set(value), [value]);

  React.useMemo(() => {
    // 每次 render 把 value 值对应的 options 存进 selectedCache
    for (const o of valueSet) {
      const option = optionsMap.get(o);
      const cache =
        option || // 优先使用当前 options 里面的数据
        selectedCache.get(o) || // 其次如果之前缓存过，仍使用之前缓存的数据
        ({ value: o } as OP); // 最后，用 value 构造一个 option
      selectedCache.set(o, cache);
    }
  }, [optionsMap, selectedCache, valueSet]);
  /**
   * 已选选项数据
   */
  const selectedOptions = React.useMemo(() => {
    const selected: OP[] = [];
    valueSet.forEach(v => {
      const cached = selectedCache.get(v);
      if (cached) {
        selected.push(cached);
      } else {
        // 对异常的处理：可能存在 value 不在 option 里面的情况
        selected.push({
          value: v,
        } as OP);
      }
    });
    return selected;
  }, [selectedCache, valueSet]);

  const noneDisableKeys = React.useMemo(
    () => filteredOptions.filter(option => !option.disabled).map(o => o.value),
    [filteredOptions],
  );

  const setSearchValue = React.useCallback(
    (v: string) => {
      onSearch?.(v);
      _setSearchValue(v);
    },
    [onSearch],
  );

  /**
   * 反选
   */
  const toggle = React.useCallback(
    (v: V) => {
      let newValue: V[];
      /**
       * 可能在删除已选值时，已选值不在 options 里面
       * 这里要从 optionsMap、selectedCache 分别尝试取一下
       **/
      const option = optionsMap.get(v) || selectedCache.get(v);

      if (!option) {
        return;
      }

      if (valueSet.has(v)) {
        newValue = value.filter(k => k !== v);
        onDeSelect?.(v, { option });
        setValue(newValue, TransferChangeAction.DeSelect);
      } else {
        onSelect?.(v, { option });
        newValue = [...value, v];
        setValue(newValue, TransferChangeAction.Select);
      }
    },
    [optionsMap, valueSet, value, onDeSelect, setValue, onSelect, selectedCache],
  );

  const isAllSelected = React.useMemo(
    () => noneDisableKeys.every(key => valueSet.has(key)) && noneDisableKeys.length > 0,
    [noneDisableKeys, valueSet],
  );

  /**
   * 清除所有
   */
  const clearAll = React.useCallback(() => {
    onClear?.();
    setValue(
      value.filter(ele => optionsMap.get(ele)?.disabled),
      TransferChangeAction.ClearAll,
    );
  }, [onClear, optionsMap, setValue, value]);

  /**
   * 反选所有
   */
  const toggleSelectAll = React.useCallback(() => {
    if (isAllSelected) {
      onToggleSelectAll?.(false);
      setValue(
        value.filter(ele => !noneDisableKeys.includes(ele)),
        TransferChangeAction.DeSelectAll,
      );
    } else {
      onToggleSelectAll?.(true);

      noneDisableKeys.forEach(v => {
        valueSet.add(v);
      });
      setValue([...valueSet], TransferChangeAction.SelectAll);
    }
  }, [isAllSelected, noneDisableKeys, onToggleSelectAll, setValue, value, valueSet]);

  return {
    value,
    setValue,
    selectedOptions,
    toggleSelectAll,
    cssScope,
    toggle,
    searchValue,
    setSearchValue,
    valueSet,
    isAllSelected,
    clearAll,
    filteredOptions,
    groupWeakMap,
    group,
  };
}

const { useContext: _useTransferContext, Provider: _TransferProvider } = createContextHook<TransferContextType>(
  'TransferContext',
  {
    strict: true,
  },
);

export const useTransferContext = _useTransferContext as <
  V extends TransferValue = string,
  OP extends TransferOption<V> = TransferOption<V>,
>() => TransferContextType<V, OP>;
export const TransferProvider = _TransferProvider as <
  V extends TransferValue = string,
  OP extends TransferOption<V> = TransferOption<V>,
>(props: {
  value: TransferContextType<V, OP>;
  children: React.ReactNode;
}) => React.ReactElement;
