// @universe-design/react/src/transfer/hooks/TransferOptionContext.ts
import type React from 'react';
import { createContextHook } from '@universe-design/react-utilities';
import type { TransferOption, TransferValue } from '../interface';
import type { TransferContextType } from './TransferContext';

export type TransferOptionContextType<
  V extends TransferValue = string,
  OP extends TransferOption<V> = TransferOption<V>,
> = {
  /**
   * 选项详情
   */
  option: OP;
  /**
   * 是否已经被选中
   */
  isChecked: boolean;
  /**
   * 是否处于禁用态
   */
  disabled: boolean;
} & Pick<TransferContextType<V, OP>, 'searchValue' | 'toggle'>;

const { useContext: _useTransferOptionContext, Provider: _TransferOptionContextProvider } =
  createContextHook<TransferOptionContextType>('TransferContext', {
    strict: true,
  });

export const useTransferOptionContext = _useTransferOptionContext as <
  V extends TransferValue = string,
  OP extends TransferOption<V> = TransferOption<V>,
>() => TransferOptionContextType<V, OP>;
export const TransferOptionContextProvider = _TransferOptionContextProvider as <
  V extends TransferValue = string,
  OP extends TransferOption<V> = TransferOption<V>,
>(props: {
  value: TransferOptionContextType<V, OP>;
  children: React.ReactNode;
}) => React.ReactElement;
