// @universe-design/react/src/transfer/hooks/useTransferData.ts
import * as React from 'react';
import { useConstant } from '@universe-design/react-utilities';
import { devWarning, warning } from '../../util';
import type { TransferGroupOption, TransferOption, TransferValue, UseTransferProps } from '../interface';

export const useTransferData = <V extends TransferValue, OP extends TransferOption<V>>({
  options,
  searchValue,
  filterOption,
  group: groupProp,
}: Required<Pick<UseTransferProps<V, OP>, 'options' | 'value' | 'filterOption' | 'group'>> & {
  searchValue: string;
}) => {
  const group = groupProp !== false;

  const groupWeakMap: WeakMap<OP, TransferGroupOption<V, OP>> = useConstant(() => new WeakMap());

  const { formattedOptions, optionsMap } = React.useMemo(() => {
    const _formattedOptions: OP[] = [];
    const _optionsMap: Map<V, OP> = new Map();

    /**
     * 用于检测是否存在 value 重复
     */
    const optionValueSet = new Set<V>();
    /**
     * 用于检测是否存在 value 重复
     */
    const groupKeySet = new Set<string>();
    function addToOptionList(op: OP) {
      const { value } = op;
      // 检测是否重复
      warning(!optionValueSet.has(value), devWarning('transfer', `option's value cannot be repeated: ${value}`));

      optionValueSet.add(value);
      _formattedOptions.push(op);
      _optionsMap.set(value, op);
    }

    options.forEach(option => {
      if (group) {
        const groupOption = option as TransferGroupOption<V, OP>;
        const { key } = groupOption;
        // 检测是否重复
        warning(!groupKeySet.has(key), devWarning('transfer', `group's key cannot be repeated: ${key}`));

        groupKeySet.add(key);

        groupOption.children.forEach(op => {
          groupWeakMap.set(op, groupOption);
          addToOptionList(op);
        });
      } else {
        addToOptionList(option as OP);
      }
    });

    return {
      formattedOptions: _formattedOptions,
      optionsMap: _optionsMap,
    };
  }, [group, groupWeakMap, options]);

  /**
   * 根据搜索条件筛选后的值
   */
  const filteredOptions = React.useMemo(
    () => formattedOptions.filter(op => filterOption(searchValue, op)),
    [filterOption, formattedOptions, searchValue],
  );

  return {
    formattedOptions,
    optionsMap,
    filteredOptions,
    groupWeakMap,
  };
};
