// @universe-design/react/src/details-display/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Vertical from '../demo/0.vertical';
import C1Align from '../demo/1.align';
import C2Column from '../demo/2.column';
import C3Size from '../demo/3.size';
import C4Custom from '../demo/4.custom';
if (!(C0Vertical as Story).storyName) {
  (C0Vertical as Story).storyName = '0. 垂直布局';
  (C0Vertical as Story).parameters = {
    docs: {
      description: {
        story: '垂直方向的详情展示\n\n',
      },
    },
  };
}
if (!(C1Align as Story).storyName) {
  (C1Align as Story).storyName = '1. 对齐方式';
  (C1Align as Story).parameters = {
    docs: {
      description: {
        story:
          "目前 `layout` 可取值为 `'horizontal' | 'vertical' | 'inline'`。 `align` 可取值为 `'center' | 'justify' | 'start'`。\n\n在代码底层实现上：\n\n1. 当 `layout = 'horizontal'`，`align` 可取值为 `'center' | 'justify' | 'start'`；\n2. 当 `layout = 'vertical'`，`align` 可取值为 `'start'`；\n3. 当 `layout = 'inline'`，`align` 可取值为 `'start'`。\n\n**UD 标准设计规范建议 `align= 'start'`为最佳！**\n\n",
      },
    },
  };
}
if (!(C2Column as Story).storyName) {
  (C2Column as Story).storyName = '2. 多列布局';
  (C2Column as Story).parameters = {
    docs: {
      description: {
        story: '单列布局、双列布局、三列布局。\n\n',
      },
    },
  };
}
if (!(C3Size as Story).storyName) {
  (C3Size as Story).storyName = '3. 不同尺寸';
  (C3Size as Story).parameters = {
    docs: {
      description: {
        story: '不同尺寸用法。\n\n',
      },
    },
  };
}
if (!(C4Custom as Story).storyName) {
  (C4Custom as Story).storyName = '4. 自定义 dom 样式';
  (C4Custom as Story).parameters = {
    docs: {
      description: {
        story: '通过自定义 ReactNode、style、className 展示不同样式的详情展示组件。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/DetailsDisplay/官网示例',
  tags: ['autodocs'],
};

export { C0Vertical, C1Align, C2Column, C3Size, C4Custom };
