// @universe-design/react/lib/dialog/Confirm.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ConfirmContent = exports.Confirm = void 0;
exports.withConfirm = withConfirm;
exports.withError = withError;
exports.withInfo = withInfo;
exports.withSuccess = withSuccess;
exports.withWarn = withWarn;
var _uwejdxd = _interopRequireDefault(require("./../.stylex/uwejdxd"));
var _u1purjcd = _interopRequireDefault(require("./../.stylex/u1purjcd"));
var _uuhcnxi = _interopRequireDefault(require("./../.stylex/uuhcnxi"));
var _u2o5q4q = _interopRequireDefault(require("./../.stylex/u2o5q4q"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _ErrorFilled = _interopRequireDefault(require("@universe-design/icons-react/ErrorFilled"));
var _InfoFilled = _interopRequireDefault(require("@universe-design/icons-react/InfoFilled"));
var _SucceedFilled = _interopRequireDefault(require("@universe-design/icons-react/SucceedFilled"));
var _WarningFilled = _interopRequireDefault(require("@universe-design/icons-react/WarningFilled"));
var _clsx = _interopRequireDefault(require("clsx"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _stylex2 = require("../stylex");
var _sx = require("./sx");
var _dialog = require("./dialog");
var _button = require("../button");
var _hooks = require("../config-provider/hooks");
var _util = require("../util");
var _hooks2 = require("../hooks");
var _FirstLineAlign = require("../shared/components/FirstLineAlign");
require("./style");
//.ugr6219{margin-top:auto;}
//.u11qmguv{margin-right:0;}
//.u10jk5vf{margin-bottom:auto;}
//.u1tyq0we{margin-left:0;}
/**
 * @public
 * @docgen
 */
const ConfirmContent = props => {
  const {
    icon: iconProp,
    title,
    content,
    cancelText,
    okText,
    cancelButtonProps,
    okButtonProps,
    enableCancelButton = true,
    autoFocusButton = 'ok',
    bodyStyle,
    onCancel,
    onOk,
    onClose,
    type: typeProp,
    extraAction,
    // style
    style,
    className,
    // ref
    okButtonRef,
    cancelButtonRef,
    sx = []
  } = props;
  const type = typeProp !== null && typeProp !== void 0 ? typeProp : 'confirm';
  const icon = (() => {
    /**
     * 现实效果: 用户传入icon: null时 icon不显示 (当用户不穿入icon属性时 icon为undefined)
     * 使用isNonNullable函数 -> iconProp为 null 与 undefined 时均根据typeProp选择相应icon
     * 此处对用户不传iconProp 和 传入iconProp: null分开处理
     */
    if (iconProp !== undefined) {
      return iconProp;
    }
    switch (typeProp) {
      case 'error':
        return /*#__PURE__*/_react.default.createElement(_ErrorFilled.default, null);
      case 'success':
        return /*#__PURE__*/_react.default.createElement(_SucceedFilled.default, null);
      case 'info':
        return /*#__PURE__*/_react.default.createElement(_InfoFilled.default, null);
      case 'confirm':
      case 'warning':
        return /*#__PURE__*/_react.default.createElement(_WarningFilled.default, null);
      default:
        return null;
    }
  })();
  const cssScope = (0, _hooks.useCssScope)('confirm');
  const locale = (0, _hooks.useLocale)();
  const cancelRef = _react.default.useRef(null);
  _react.default.useImperativeHandle(cancelButtonRef, () => cancelRef.current);
  const okRef = _react.default.useRef(null);
  _react.default.useImperativeHandle(okButtonRef, () => okRef.current);
  const [okLoading, setOkLoading] = (0, _react.useState)(Boolean(okButtonProps === null || okButtonProps === void 0 ? void 0 : okButtonProps.loading));
  const [cancelLoading, setCancelLoading] = (0, _react.useState)(Boolean(cancelButtonProps === null || cancelButtonProps === void 0 ? void 0 : cancelButtonProps.loading));
  const triggerClose = () => onClose && onClose();

  // 自动设置焦点，回车后会触发相关事件
  (0, _hooks2.useEffectOnce)(() => {
    let timerId;
    if (autoFocusButton) {
      timerId = window.setTimeout(() => {
        var _cancelRef$current, _okRef$current;
        autoFocusButton === 'cancel' ? (_cancelRef$current = cancelRef.current) === null || _cancelRef$current === void 0 ? void 0 : _cancelRef$current.focus() : (_okRef$current = okRef.current) === null || _okRef$current === void 0 ? void 0 : _okRef$current.focus();
      });
    }
    return () => {
      if (timerId) {
        window.clearTimeout(timerId);
      }
    };
  });
  (0, _react.useEffect)(() => {
    setOkLoading(Boolean(okButtonProps === null || okButtonProps === void 0 ? void 0 : okButtonProps.loading));
    setCancelLoading(Boolean(cancelButtonProps === null || cancelButtonProps === void 0 ? void 0 : cancelButtonProps.loading));
  }, [okButtonProps, cancelButtonProps]);
  const handleOnConfirmClick = event => {
    if (okButtonProps && okButtonProps.onClick) {
      okButtonProps.onClick(event);
    }
    if (!onOk) {
      triggerClose();
    } else {
      if (onOk.length) {
        onOk(triggerClose);
      } else {
        const result = onOk();
        if (result && result.then) {
          setOkLoading(true);
          result.then(() => {
            setOkLoading(false);
            triggerClose();
          }, e => {
            console.error(e);
            setOkLoading(false);
          });
        } else {
          triggerClose();
        }
      }
    }
  };
  const handleOnCancelClick = event => {
    if (cancelButtonProps && cancelButtonProps.onClick) {
      cancelButtonProps.onClick(event);
    }
    if (!onCancel) {
      triggerClose();
    } else {
      if (onCancel.length) {
        onCancel(triggerClose);
      } else {
        const result = onCancel();
        if (result && result.then) {
          setCancelLoading(true);
          result.then(() => {
            setCancelLoading(false);
            triggerClose();
          }, e => {
            console.error(e);
            setCancelLoading(false);
          });
        } else {
          triggerClose();
        }
      }
    }
  };
  const headerNode = /*#__PURE__*/_react.default.createElement("div", {
    className: cssScope('__header')
  }, /*#__PURE__*/_react.default.createElement(_FirstLineAlign.FirstLineAlign, {
    className: cssScope('__title'),
    sx: [_stylex2.cssTypographyTitle4],
    beforeProps: (0, _util.isNonNullable)(icon) ? {
      className: cssScope('__icon'),
      sx: [_sx.sxDialogIconWrapper],
      children: /*#__PURE__*/_react.default.createElement("div", {
        className: (0, _stylex.default)(_sx.sxDialogIcon)
      }, icon)
    } : undefined
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: cssScope('__titleContent')
  }, title)), (0, _util.isNonNullable)(content) ? /*#__PURE__*/_react.default.createElement("div", {
    style: bodyStyle,
    className: (0, _clsx.default)(cssScope('__body'), (0, _stylex.default)(_stylex2.cssTypographyBody0))
  }, content) : null);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(className, cssScope('__content'), {
      [cssScope('-hasIcon')]: icon,
      [cssScope(`-${type}`)]: true
    }, (0, _stylex.default)(_sx.sxModalContent.root, _stylex2.cssTypographyBody0, ...sx)),
    style: style
  }, headerNode, (() => /*#__PURE__*/_react.default.createElement("div", {
    className: cssScope('__footer')
  }, extraAction ? /*#__PURE__*/_react.default.createElement("div", {
    className: cssScope('__footer__extraAction')
  }, extraAction) : null, /*#__PURE__*/_react.default.createElement("div", {
    className: cssScope('__footer__btns')
  }, /*#__PURE__*/_react.default.createElement(_button.Button, (0, _extends2.default)({
    ref: okRef,
    type: "primary"
  }, okButtonProps, {
    onClick: handleOnConfirmClick,
    loading: okLoading
  }), okText || locale.components.modal.okText), enableCancelButton && /*#__PURE__*/_react.default.createElement(_button.Button, (0, _extends2.default)({
    ref: cancelRef
  }, cancelButtonProps, {
    loading: cancelLoading,
    onClick: handleOnCancelClick
  }), cancelText || locale.components.modal.cancelText))))());
};

// eslint-disable-next-line max-lines-per-function
exports.ConfirmContent = ConfirmContent;
const Confirm = props => {
  const {
    icon,
    title,
    content,
    keyboardClosable = true,
    mask = true,
    centered = true,
    destroyOnClose,
    visible,
    cancelText,
    okText,
    cancelButtonProps,
    okButtonProps,
    enableCancelButton = true,
    maskStyle,
    bodyStyle,
    autoFocusButton = 'ok',
    onCancel,
    onOk,
    onClose,
    // 内部属性
    onAfterOpen,
    onAfterClose,
    className,
    style,
    wrapClassName,
    type = 'confirm',
    extraAction,
    maskClosable = false,
    ...restProps
  } = props;
  const cssScope = (0, _hooks.useCssScope)('confirm');
  const triggerClose = () => onClose && onClose();
  const initialFocusRef = _react.default.useRef(null);
  return /*#__PURE__*/_react.default.createElement(_dialog.Dialog, (0, _extends2.default)({
    visible: visible,
    centered: centered,
    onClose: () => {
      triggerClose();
      (0, _util.isFunction)(onCancel) && onCancel();
    },
    overlay: mask,
    onAfterOpen: onAfterOpen,
    onAfterClose: () => {
      (0, _util.isFunction)(onAfterClose) && onAfterClose();
    },
    destroyOnClose: destroyOnClose,
    overlayStyle: maskStyle,
    wrapperClassName: wrapClassName,
    contentStyle: style,
    overlayClosable: maskClosable,
    keyboardClosable: keyboardClosable,
    contentClassName: (0, _clsx.default)(cssScope(), className),
    size: "small",
    initialFocusRef: autoFocusButton ? initialFocusRef : undefined
  }, restProps), /*#__PURE__*/_react.default.createElement(ConfirmContent, {
    icon,
    title,
    content,
    cancelText,
    okText,
    cancelButtonProps,
    okButtonProps,
    enableCancelButton,
    autoFocusButton,
    bodyStyle,
    onCancel,
    onOk,
    onClose,
    type,
    extraAction,
    sx: [centered && {
      "props": {
        "B6of3ja": _uwejdxd.default,
        "t21cq0": _u1purjcd.default,
        "jrapky": _uuhcnxi.default,
        "Frg6f3": _u2o5q4q.default
      }
    }],
    okButtonRef: autoFocusButton === 'ok' ? initialFocusRef : undefined,
    cancelButtonRef: autoFocusButton === 'cancel' ? initialFocusRef : undefined
  }));
};
exports.Confirm = Confirm;
function withWarn(props) {
  return {
    enableCancelButton: false,
    ...props,
    type: 'warning'
  };
}
function withInfo(props) {
  return {
    enableCancelButton: false,
    ...props,
    type: 'info'
  };
}
function withSuccess(props) {
  return {
    enableCancelButton: false,
    ...props,
    type: 'success'
  };
}
function withError(props) {
  return {
    enableCancelButton: false,
    ...props,
    type: 'error'
  };
}
function withConfirm(props) {
  return {
    enableCancelButton: true,
    ...props,
    type: 'confirm'
  };
}
