// @universe-design/react/lib/dialog/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Confirm", {
  enumerable: true,
  get: function () {
    return _Confirm.Confirm;
  }
});
Object.defineProperty(exports, "ConfirmContent", {
  enumerable: true,
  get: function () {
    return _Confirm.ConfirmContent;
  }
});
exports.Dialog = void 0;
Object.defineProperty(exports, "DialogCancelType", {
  enumerable: true,
  get: function () {
    return _dialog.DialogCancelType;
  }
});
Object.defineProperty(exports, "DialogLayoutMode", {
  enumerable: true,
  get: function () {
    return _dialog.DialogLayoutMode;
  }
});
Object.defineProperty(exports, "useDialog", {
  enumerable: true,
  get: function () {
    return _useDialog.useDialog;
  }
});
var _react = _interopRequireDefault(require("react"));
var _reactDom = _interopRequireDefault(require("react-dom"));
var _Confirm = require("./Confirm");
var _Modal = require("./Modal");
var _util = require("../util");
var _configProvider = require("../config-provider");
var _useDialog = require("./hooks/useDialog");
var _dialog = require("./dialog");
const destroyMap = new Map();
const Dialog = exports.Dialog = _Modal.Modal;
function confirm(_ref) {
  let {
    onAfterClose,
    visible,
    ...config
  } = _ref;
  const uuid = `ud-confirm-instance-${new Date().getTime()}`;
  const container = document.createElement('div');
  document.body.appendChild(container);
  let currentConfig;
  function destroy() {
    const unmountResult = _reactDom.default.unmountComponentAtNode(container);
    if (unmountResult && container.parentNode) {
      container.parentNode.removeChild(container);
    }
    // config.onCancel && config.onCancel();
    destroyMap.delete(uuid);
  }
  function render(renderConfig) {
    _reactDom.default.render(/*#__PURE__*/_react.default.createElement(_configProvider.CrossRootConfigProvider, null, /*#__PURE__*/_react.default.createElement(_Confirm.Confirm, renderConfig)), container);
  }
  function close() {
    currentConfig = {
      ...currentConfig,
      visible: false
    };
    render(currentConfig);
  }
  function update(newConfig) {
    if ((0, _util.isFunction)(newConfig)) {
      currentConfig = newConfig(currentConfig);
    } else {
      currentConfig = {
        ...currentConfig,
        ...newConfig
      };
    }
    render(currentConfig);
  }
  currentConfig = {
    ...config,
    onAfterClose: () => {
      (0, _util.isFunction)(onAfterClose) && onAfterClose();
      destroy();
    },
    visible: true,
    onClose: close
  };
  render(currentConfig);
  destroyMap.set(uuid, destroy);
  return {
    destroy: close,
    update
  };
}
Dialog.destroyAll = () => {
  destroyMap.forEach(fn => fn());
  destroyMap.clear();
};
Dialog.Content = _Modal.ModalContent;
Dialog.confirm = config => confirm((0, _Confirm.withConfirm)(config));
Dialog.info = config => confirm((0, _Confirm.withInfo)(config));
Dialog.success = config => confirm((0, _Confirm.withSuccess)(config));
Dialog.error = config => confirm((0, _Confirm.withError)(config));
Dialog.warning = config => confirm((0, _Confirm.withWarn)(config));
