// @universe-design/react/lib/dialog/Modal.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ModalContent = exports.Modal = void 0;
var _um9xw6y = _interopRequireDefault(require("./../.stylex/um9xw6y"));
var _uwejdxd = _interopRequireDefault(require("./../.stylex/uwejdxd"));
var _u1purjcd = _interopRequireDefault(require("./../.stylex/u1purjcd"));
var _uuhcnxi = _interopRequireDefault(require("./../.stylex/uuhcnxi"));
var _u2o5q4q = _interopRequireDefault(require("./../.stylex/u2o5q4q"));
var _ux645fm = _interopRequireDefault(require("./../.stylex/ux645fm"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _stylex2 = require("../stylex");
var _dialog = require("./dialog");
var _button = require("../button");
var _CloseOutlined = _interopRequireDefault(require("@universe-design/icons-react/CloseOutlined"));
var _hooks = require("../config-provider/hooks");
var _hooks2 = require("../hooks");
var _reactUtilities = require("@universe-design/react-utilities");
var _util = require("../util");
var _getThunkElement = require("../util/getThunkElement");
var _FirstLineAlign = require("../shared/components/FirstLineAlign");
var _sx = require("./sx");
require("./style");
//.uwegl26{box-shadow:@shadow-s5-down;}
//.ugr6219{margin-top:auto;}
//.u11qmguv{margin-right:0;}
//.u10jk5vf{margin-bottom:auto;}
//.u1tyq0we{margin-left:0;}
//.u1a3p1vp{overflow:hidden;}
/* eslint-disable @typescript-eslint/no-use-before-define */
function useCheckVerticalOverflow(_ref) {
  let {
    target
  } = _ref;
  const [overflowTop, setOverflowTop] = (0, _react.useState)(false);
  const [overflowBottom, setOverflowBottom] = (0, _react.useState)(false);
  const checkVerticalOverflow = (0, _react.useCallback)(() => {
    const targetEl = (0, _getThunkElement.getThunkElement)(target);
    if (!targetEl) {
      return;
    }
    const {
      scrollTop,
      scrollHeight,
      clientHeight
    } = targetEl;
    const targetInset = {
      top: 0,
      bottom: clientHeight
    };
    const childInset = {
      top: -scrollTop,
      bottom: -scrollTop + scrollHeight
    };
    if (targetInset.top > childInset.top) {
      setOverflowTop(true);
    } else {
      setOverflowTop(false);
    }
    if (targetInset.bottom < childInset.bottom) {
      setOverflowBottom(true);
    } else {
      setOverflowBottom(false);
    }
  }, [target]);
  return {
    overflow: {
      top: overflowTop,
      bottom: overflowBottom
    },
    checkVerticalOverflow
  };
}

/**
 * @public
 * @docgen
 * <Dialog.Content /> 的属性。
 */

/**
 * @public
 * @docgen
 * <Dialog /> 的属性。
 */

const ModalContent = exports.ModalContent = /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
  const cssScope = (0, _hooks.useCssScope)('modal');
  const {
    header,
    title,
    description,
    closable = true,
    enableCancelButton = true,
    children,
    cancelText,
    okText,
    cancelButtonProps,
    okButtonProps,
    bodyStyle = {},
    confirmLoading,
    footer,
    onCancel,
    onOk,
    extraAction,
    style,
    className,
    sx = [],
    bodyHeightResponsiveMode = 'scroll',
    maxHeight,
    ...restProps
  } = props;
  const locale = (0, _hooks.useLocale)();
  const layoutMode = (0, _dialog.useDialogLayoutMode)();
  const closeIconNode = closable ? /*#__PURE__*/_react.default.createElement(_button.Button, {
    className: cssScope('__close'),
    onClick: e => {
      (0, _util.isFunction)(onCancel) && onCancel(e, {
        cancelType: _dialog.DialogCancelType.CloseIcon
      });
    },
    icon: /*#__PURE__*/_react.default.createElement(_CloseOutlined.default, null),
    type: "icon"
  }) : null;
  const bodyScrollable = bodyHeightResponsiveMode === 'scroll';
  const bodyStretch = bodyHeightResponsiveMode === 'stretch';
  const [headerSizeRef, headerSize] = (0, _hooks2.useSize)({
    resizeObserverDisabled: !bodyScrollable && !bodyStretch
  });
  const [footerSizeRef, footerSize] = (0, _hooks2.useSize)({
    resizeObserverDisabled: !bodyScrollable && !bodyStretch
  });
  const contentCls = (0, _clsx.default)(className, cssScope('__content'), closeIconNode && cssScope('-closable'), (0, _stylex.default)(_sx.sxModalContent.root, bodyScrollable && _sx.sxModalContent.scrollable, _stylex2.cssTypographyBody0, bodyHeightResponsiveMode === 'stretch' && _sx.sxModalContent.stretch, ...sx));
  const handleCancel = e => {
    if (cancelButtonProps && cancelButtonProps.onClick) {
      cancelButtonProps.onClick(e);
    }
    (0, _util.isFunction)(onCancel) && onCancel(e, {
      cancelType: _dialog.DialogCancelType.CancelButton
    });
  };
  const handleOk = e => {
    if (okButtonProps && okButtonProps.onClick) {
      okButtonProps.onClick(e);
    }
    (0, _util.isFunction)(onOk) && onOk(e);
  };

  /**
   * Title ReactNode
   */
  let headerNode = null;
  if (header !== null) {
    headerNode = /*#__PURE__*/_react.default.createElement("div", {
      ref: headerSizeRef,
      className: cssScope('__header')
    }, (0, _util.isNonNullable)(header) ? header : /*#__PURE__*/_react.default.createElement(_FirstLineAlign.FirstLineAlign, {
      className: cssScope('__title'),
      sx: [_stylex2.cssTypographyTitle4]
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: cssScope('__titleContent')
    }, title)), (0, _util.isNonNullable)(description) ? /*#__PURE__*/_react.default.createElement("div", {
      className: (0, _clsx.default)(cssScope('__des'), (0, _stylex.default)(_stylex2.cssTypographyBody0))
    }, description) : null);
  }

  /**
   * Footer ReactNode
   */
  const footerNode = (() => {
    const footerContentNode = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, extraAction ? /*#__PURE__*/_react.default.createElement("div", {
      className: cssScope('__footer__extraAction')
    }, extraAction) : null, /*#__PURE__*/_react.default.createElement("div", {
      className: cssScope('__footer__btns')
    }, /*#__PURE__*/_react.default.createElement(_button.Button, (0, _extends2.default)({
      type: "primary"
    }, okButtonProps, {
      loading: confirmLoading || (okButtonProps === null || okButtonProps === void 0 ? void 0 : okButtonProps.loading),
      onClick: handleOk
    }), okText || locale.components.modal.okText), enableCancelButton && /*#__PURE__*/_react.default.createElement(_button.Button, (0, _extends2.default)({}, cancelButtonProps, {
      onClick: handleCancel
    }), cancelText || locale.components.modal.cancelText)));
    if (footer) {
      return /*#__PURE__*/_react.default.createElement("div", {
        ref: footerSizeRef,
        className: cssScope('__footer')
      }, footer);
    }
    if (footer === null) {
      return null;
    }
    return /*#__PURE__*/_react.default.createElement("div", {
      ref: footerSizeRef,
      className: cssScope('__footer')
    }, footerContentNode);
  })();

  /**
   * Body ReactNode
   */

  // body 内容超出增加分割线
  const bodyRef = (0, _reactUtilities.useResizeObserver)(() => {
    checkVerticalOverflow();
  });
  const {
    overflow: bodyOverflow,
    checkVerticalOverflow
  } = useCheckVerticalOverflow({
    target: bodyRef
  });
  const childrenRef = (0, _reactUtilities.useResizeObserver)(() => {
    checkVerticalOverflow();
  });
  (0, _react.useEffect)(() => {
    const body = bodyRef.current;
    if (!body) {
      return;
    }
    body.addEventListener('scroll', checkVerticalOverflow);
    return () => {
      body.removeEventListener('scroll', checkVerticalOverflow);
    };
  }, [bodyRef, checkVerticalOverflow]);
  const childrenMaybeWithRef = (0, _reactUtilities.useCloneElementRef)({
    children,
    ref: childrenRef
  });
  const {
    height: bodyHeight
  } = bodyStyle;
  const bodyMinHeight = (() => {
    if ((0, _util.isNonNullable)(bodyStyle.minHeight)) {
      return bodyStyle.minHeight;
    }
    if (layoutMode === 'legacy') {
      return 0;
    }
    // 62 是 body 的最小高度，一个字段 32 + 8 + 22
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    return 32 + 8 + 22;
  })();
  const bodyNode = (0, _util.isNonNullable)(children) ? /*#__PURE__*/_react.default.createElement('div', {
    style: {
      ...bodyStyle,
      minHeight: bodyMinHeight,
      overflowY: bodyScrollable ? 'auto' : undefined
    },
    className: (0, _clsx.default)(cssScope('__body'), headerNode && bodyOverflow.top && cssScope('__body--top-overflow'), footerNode && bodyOverflow.bottom && cssScope('__body--bottom-overflow'), (0, _stylex.default)(bodyStretch && {
      "props": {
        "Bpg54ce": _ux645fm.default
      }
    })),
    ref: bodyRef
  }, childrenMaybeWithRef) : null;
  const contentStyle = (() => {
    // 24 目前在 body 底部的 margin

    const sharedStyle = {
      ...style,
      maxHeight
    };
    if (bodyScrollable || bodyStretch) {
      const bodyMinHeightStr = (() => {
        if (!bodyNode) {
          return;
        }
        if ((0, _util.isNonNullable)(bodyHeight)) {
          return typeof bodyHeight === 'number' ? `${bodyHeight}px` : bodyHeight;
        }
        return typeof bodyMinHeight === 'number' ? `${bodyMinHeight}px` : bodyMinHeight;
      })();
      const minHeight = `calc(${[`${headerSize.height + footerSize.height + (bodyNode ? 24 : 0)}px`, bodyMinHeightStr].filter(Boolean).join(' + ')})`;
      return {
        ...sharedStyle,
        minHeight
      };
    }
    return sharedStyle;
  })();
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
    className: contentCls,
    style: contentStyle,
    ref: ref
  }, restProps), headerNode, closeIconNode, bodyNode, footerNode);
});

// eslint-disable-next-line max-lines-per-function
const Modal = exports.Modal = /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
  const {
    title,
    zIndex,
    description,
    closable = true,
    enableCancelButton = true,
    maskClosable = true,
    keyboardClosable = true,
    centered = true,
    width,
    visible,
    children,
    cancelText,
    okText,
    cancelButtonProps,
    okButtonProps,
    mask = true,
    style,
    maskStyle,
    bodyStyle,
    enableAnimation,
    transitionName,
    maskTransitionName,
    confirmLoading,
    // SSR compatiable
    getContainer,
    footer,
    onCancel,
    onOk,
    onAfterOpen,
    onAfterClose,
    className,
    wrapClassName,
    extraAction,
    size = 'medium',
    header,
    safeMargin,
    onClose,
    bodyHeightResponsiveMode,
    maxHeight,
    ...restProps
  } = props;
  const cssScope = (0, _hooks.useCssScope)('modal');
  const layoutMode = (0, _dialog.useDialogLayoutMode)();
  return /*#__PURE__*/_react.default.createElement(_dialog.Dialog, (0, _extends2.default)({
    ref: ref,
    width: width,
    zIndex: zIndex,
    visible: visible,
    centered: centered,
    overlay: mask,
    overlayStyle: maskStyle,
    wrapperClassName: wrapClassName,
    overlayClosable: maskClosable,
    contentStyle: style,
    contentClassName: (0, _clsx.default)(cssScope(), className),
    keyboardClosable: keyboardClosable,
    enableAnimation: enableAnimation,
    overlayTransitionClassName: transitionName,
    contentTransitionClassName: maskTransitionName,
    getContainer: getContainer,
    onClose: (e, payload) => {
      (0, _util.isFunction)(onCancel) && onCancel(e, payload);
      if (e !== null && e !== void 0 && e.isDefaultPrevented()) {
        return;
      }
      // 这个逻辑只给 imperative dialog 使用
      (0, _util.isFunction)(onClose) && onClose();
    },
    onAfterOpen: onAfterOpen,
    onAfterClose: onAfterClose,
    size: size,
    safeMargin: safeMargin
  }, restProps), /*#__PURE__*/_react.default.createElement(ModalContent, {
    header,
    title,
    description,
    closable,
    enableCancelButton,
    children,
    cancelText,
    okText,
    cancelButtonProps,
    okButtonProps,
    bodyStyle,
    confirmLoading,
    footer,
    onCancel,
    onOk,
    extraAction,
    sx: [layoutMode === 'legacy' ? [] : [centered && {
      "props": {
        "B6of3ja": _uwejdxd.default,
        "t21cq0": _u1purjcd.default,
        "jrapky": _uuhcnxi.default,
        "Frg6f3": _u2o5q4q.default
      }
    }, !mask && {
      "props": {
        "E5pizo": _um9xw6y.default
      }
    }]],
    bodyHeightResponsiveMode,
    maxHeight
  }));
});
