// @universe-design/react/lib/dialog/sx.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.sxModalContent = exports.sxDialogIconWrapper = exports.sxDialogIcon = void 0;
var _udg2fzc = _interopRequireDefault(require("./../.stylex/udg2fzc"));
var _u1j86i = _interopRequireDefault(require("./../.stylex/u1j86i34"));
var _u5jxv = _interopRequireDefault(require("./../.stylex/u5jxv71"));
var _uohwdt = _interopRequireDefault(require("./../.stylex/uohwdt"));
var _u13d3rh = _interopRequireDefault(require("./../.stylex/u13d3rh8"));
var _udavndu = _interopRequireDefault(require("./../.stylex/udavndu"));
var _u5p8ukz = _interopRequireDefault(require("./../.stylex/u5p8ukz"));
var _u1mtnbey = _interopRequireDefault(require("./../.stylex/u1mtnbey"));
var _u1hcffm = _interopRequireDefault(require("./../.stylex/u1hcffm1"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _stylex2 = require("../stylex");
//.u1l02sjl{height:100%;}
//.u1immsc2{max-height:100%;}
//.u10pi13n{position:relative;}
//.u1vx9l62{flex-direction:column;}
//.uly5x3f{width:100%;}
//.u1rt2boy{font-size:24px;}
//.uez10in{line-height:0;}
//.u7nlbp4{align-self:center;}
//.u17vyym1{margin-right:16px;}
/**
 * tokens
 */
const DialogIconSize = 24;
const DialogIconTitleSpacing = 16;
const sxDialogIconWrapper = exports.sxDialogIconWrapper = {
  "props": {
    "t21cq0": _u1hcffm.default
  }
};
const sxDialogIcon = exports.sxDialogIcon = {
  "props": {
    "Be2twd7": _udavndu.default,
    "Bg96gwp": _u5p8ukz.default,
    "qb2dma": _u1mtnbey.default
  }
};
const sxModalContent = exports.sxModalContent = {
  root: _stylex.default.compose(_stylex2.cssFlexBase, _stylex2.sxBorderRadius.md, {
    "props": {
      "qhf8xq": _u5jxv.default,
      "Beiy3e4": _uohwdt.default,
      "a9b677": _u13d3rh.default
    }
  }),
  scrollable: {
    "props": {
      "Bxyxcbc": _u1j86i.default
    }
  },
  stretch: {
    "props": {
      "Bqenvij": _udg2fzc.default
    }
  }
};
