// @universe-design/react/src/menu/index.tsx
import stylex from '@universe-design/stylex';
import clsx from 'clsx';
import * as React from 'react';
import { cssTypographyHeadline } from '../stylex';
import type { MenuItemGroupProps } from './interface';
import type { RcMenuProps } from './internal-menu';
import type {
  RcMenuInfo,
  RenderIconType,
  MenuSelectEventHandler,
  MenuSelectInfo,
  TriggerSubMenuAction,
  RcMenuMode,
} from './internal-menu/interface';
import { Divider, RcMenuItemGroup } from './internal-menu';
import OriginMenu from './Menu';
import MenuItem from './MenuItem';
import type { TitleEventEntity } from './SubMenu';
import SubMenu from './SubMenu';

const MenuItemGroup = ({ title, ...props }: MenuItemGroupProps) => (
  <RcMenuItemGroup title={<div className={clsx(stylex(cssTypographyHeadline))}>{title}</div>} {...props} />
);

type MenuType = typeof OriginMenu & {
  Item: typeof MenuItem;
  SubMenu: typeof SubMenu;
  Divider: typeof Divider;
  ItemGroup: typeof MenuItemGroup;
} & {
  __UD_MENU?: boolean;
};

const MenuComponent = OriginMenu as MenuType;
MenuComponent.Item = MenuItem;
MenuComponent.SubMenu = SubMenu;
MenuComponent.Divider = Divider;
MenuComponent.ItemGroup = MenuItemGroup;
MenuComponent.__UD_MENU = true;

export const Menu = MenuComponent;

export type { MenuProps } from './Menu';
export type { MenuItemProps } from './MenuItem';
export type { SubMenuProps } from './SubMenu';
export type { MenuInfo, MenuMode, MenuItemGroupProps } from './interface';
export type {
  TitleEventEntity,
  RcMenuInfo,
  RenderIconType,
  MenuSelectEventHandler,
  MenuSelectInfo,
  TriggerSubMenuAction,
  RcMenuProps,
  RcMenuMode,
};
