// @universe-design/react/src/menu/MenuTitle.tsx
import stylex from '@universe-design/stylex';
import clsx from 'clsx';
import React from 'react';
import { Text } from '../text';
import { MenuContext } from './MenuContext';
import { RcMenuContext } from './internal-menu/context/MenuContext';

const sxMenuTitleContent = stylex.create({
  lineHeight: 'inherit',
  fontWeight: 'inherit',
  fontSize: 'inherit',
  flexGrow: 1,
  minWidth: 0,
});

export const MenuTitle = ({
  showEllipsis,
  title,
  style,
}: {
  showEllipsis: boolean;
  title: React.ReactNode;
  style?: React.CSSProperties;
}) => {
  const { mode: rcMode } = React.useContext(RcMenuContext);
  const { prefixCls } = React.useContext(MenuContext);

  if (showEllipsis) {
    const lineClamp = rcMode === 'vertical' || rcMode === 'inline' ? 2 : 1;

    return (
      <Text
        as="div"
        className={clsx(`${prefixCls}-item-title-content`, `${prefixCls}-item-title-content-text-overflow`)}
        style={style}
        sx={[sxMenuTitleContent]}
        lineClamp={lineClamp}
      >
        {title}
      </Text>
    );
  }
  return (
    <Text as="div" style={style} className={`${prefixCls}-item-title-content`} sx={[sxMenuTitleContent]}>
      {title}
    </Text>
  );
};
