// @universe-design/react/src/menu/Menu.tsx
import type { FC } from 'react';
import React, { useContext, useMemo } from 'react';
import clsx from 'clsx';
import RightBoldOutlined from '@universe-design/icons-react/RightBoldOutlined';
import type { RcMenuProps } from './internal-menu';
import RcMenu from './internal-menu';
import { SiderContext } from '../layout';
import { useClassPrefix, useCssScope } from '../config-provider/hooks';
import { MenuContext } from './MenuContext';
import { collapseMotion } from '../shared/motion';
import { milliSeconds2Seconds, isSubMenu } from './util';
import { devWarning, isFunction, isNonNullable, omit, warning } from '../util';
import { DropdownMenuContext } from '../dropdown/DropdownMenuContext';
import MenuMore from './MenuMore';
import { getPlacements } from './placements';
import { navMenuDefaults } from './constants';
import type { MenuMode } from './interface';
import './style';

/**
 * @docgen
 */
export type MenuProps = {
  /**
   * 菜单模式，`normal` 模式供下拉菜单使用
   */
  mode?: MenuMode;
  /**
   * `inline`时二级及以上层级菜单的缩进宽度
   *  @defaultValue 28
   */
  inlineIndent?: number;
  /**
   * 菜单项目文字是否溢出收起
   */
  textOverflow?: boolean;
  /**
   * 可选模式下拉菜单是否显示勾选标记
   */
  showCheckmark?: boolean;
} & Omit<RcMenuProps, 'mode'>;

const Menu: FC<MenuProps> = props => {
  const {
    mode: modeProp,
    inlineCollapsed = false,
    className,
    prefixCls: prefixClsInProps,
    expandIcon,
    textOverflow = modeProp !== 'normal',
    getPopupContainer = () => document.body,
    subMenuCloseDelay,
    subMenuOpenDelay,
    inlineIndent = navMenuDefaults.inlineIndent,
    showCheckmark = true,
    onClick: onClickProp,
    selectable: selectableProp,
    builtinPlacements: builtinPlacementsProp,
    overflowedIndicator: overflowedIndicatorProp,
    selectedKeys,
    defaultSelectedKeys,
  } = props;

  const rootClasPrefix = useClassPrefix();
  const siderCtx = useContext(SiderContext);
  const dropdownCtx = useContext(DropdownMenuContext);
  /**
   * 对于menu的mode属性，首先采用用户定义的mode，再考虑:
   * 如果是dropdown中的menu，即存在dropdownCtx，则mode定义为"normal"
   * 如果是普通的menu，则mode定义为"vertical"
   */
  const mergedMode = React.useMemo(
    () =>
      // eslint-disable-next-line no-nested-ternary
      isNonNullable(modeProp) ? modeProp : dropdownCtx ? 'normal' : 'vertical',
    [dropdownCtx, modeProp],
  );

  const mergedPlacements = React.useMemo(() => {
    if (builtinPlacementsProp) {
      return builtinPlacementsProp;
    }
    return getPlacements(mergedMode);
  }, [builtinPlacementsProp, mergedMode]);
  /**
   * 对于menu的selectable属性，首先采用用户定义的selectable，再考虑:
   * 如果传递了 selectedKeys, 也认为 selectable 为 true
   * 如果是dropdown中的menu，即存在dropdownCtx，则selectable定义为false
   * 如果是普通的menu，则selectable定义为true
   */
  const mergedSelectable = React.useMemo(() => {
    if (isNonNullable(selectableProp)) {
      return selectableProp;
    }

    if (isNonNullable(selectedKeys) || isNonNullable(defaultSelectedKeys)) {
      return true;
    }
    return !dropdownCtx;
  }, [defaultSelectedKeys, dropdownCtx, selectableProp, selectedKeys]);
  const cssScope = useCssScope(mergedMode === 'normal' ? 'menu-normal' : 'menu');
  const prefixCls = prefixClsInProps || cssScope();

  warning(
    // eslint-disable-next-line react/destructuring-assignment
    !(siderCtx && props.inlineCollapsed !== undefined),
    devWarning('Menu', 'prop `inlineCollapsed` will be ignored inside a <Sider /> component'),
  );

  const mergedInlineCollapsed = siderCtx ? siderCtx.siderCollapsed : inlineCollapsed;

  const { children } = props;
  const includeSubMenu = React.Children.toArray(children).find(isSubMenu);

  const menuClassName = clsx(
    className,
    dropdownCtx?.className,
    {
      [`${prefixCls}-inline-collapsed`]: mergedInlineCollapsed,
      [`${prefixCls}-include-subMenu`]: includeSubMenu,
    },
    'ud-scrollbar',
  );

  const defaultMotions = {
    horizontal: { motionName: `${rootClasPrefix}-slide-up` },
    inline: collapseMotion(rootClasPrefix),
    other: { motionName: `${rootClasPrefix}-zoom-big` },
  };

  const rcMode = mergedMode === 'normal' ? 'vertical' : mergedMode;

  const renderExpandIcon = () => (
    <span className={`${prefixCls}-submenu-arrow`}>
      {expandIcon || <RightBoldOutlined className={`${prefixCls}-submenu-arrow-icon`} />}
    </span>
  );

  return (
    <MenuContext.Provider
      value={useMemo(
        () => ({
          prefixCls,
          inlineCollapsed: mergedInlineCollapsed,
          mode: mergedMode,
          selectable: mergedSelectable,
          textOverflow,
          firstLevel: true,
          showCheckmark,
        }),
        [mergedInlineCollapsed, mergedMode, mergedSelectable, prefixCls, showCheckmark, textOverflow],
      )}
    >
      <RcMenu
        builtinPlacements={mergedPlacements}
        getPopupContainer={getPopupContainer}
        {...omit(props, ['showCheckmark', 'textOverflow'])}
        overflowedIndicator={overflowedIndicatorProp ?? <MenuMore />}
        onClick={React.useCallback<NonNullable<RcMenuProps['onClick']>>(
          e => {
            isFunction(onClickProp) && onClickProp(e);
            isFunction(dropdownCtx?.onClick) && dropdownCtx?.onClick(e);
          },
          [dropdownCtx, onClickProp],
        )}
        mode={rcMode}
        selectable={mergedSelectable}
        inlineCollapsed={mergedInlineCollapsed}
        className={menuClassName}
        prefixCls={prefixCls}
        direction="ltr"
        expandIcon={renderExpandIcon}
        defaultMotions={defaultMotions}
        {...milliSeconds2Seconds({ subMenuOpenDelay, subMenuCloseDelay })}
        inlineIndent={inlineIndent}
      />
    </MenuContext.Provider>
  );
};

export default Menu;
