// @universe-design/react/src/menu/placements.ts
import type { MenuMode } from './interface';

export function getOffset(mode: MenuMode) {
  let popupOffset;
  switch (mode) {
    case 'horizontal': {
      // 设计规范要 x 方向刚好压着边（border 重合），y 方向 menu item 要和 sub menu item 的第一个对齐
      const menuPaddingTop = 8;
      const menuItemMarginTop = 1;
      const menuBorderWidth = 1;
      const xAxiosOffset = -(menuPaddingTop + menuItemMarginTop + menuBorderWidth);
      popupOffset = [0, xAxiosOffset];
      break;
    }
    case 'vertical':
    case 'inline':
      {
        // 设计规范要 x 方向刚好压着边（border 重合），
        // y 方向 menu item 要和 sub menu item 的第一个对齐
        const menuPaddingTop = 8;
        const menuItemMarginTop = 1;
        const menuBorderWidth = 1;
        popupOffset = [8, -(menuPaddingTop + menuItemMarginTop + menuBorderWidth)];
      }
      break;
    case 'normal':
      {
        const menuPaddingTop = 2;
        const menuItemMarginTop = 1;
        const menuBorderWidth = 1;
        popupOffset = [3, -(menuPaddingTop + menuItemMarginTop + menuBorderWidth)];
      }
      break;
    default:
      break;
  }
  return popupOffset;
}

export function getPlacements(mode: MenuMode) {
  const popupOffset = getOffset(mode);
  const defaultAlign = {
    targetOffset: [0, 0],
    ignoreShake: true,
    overflow: {
      adjustX: 1,
      adjustY: 1,
      alwaysByViewport: true,
    },
  };
  return {
    topLeft: {
      points: ['bl', 'tl'],
      offset: popupOffset,
      ...defaultAlign,
    },
    bottomLeft: {
      points: ['tl', 'bl'],
      offset: popupOffset,
      ...defaultAlign,
    },
    leftTop: {
      points: ['tr', 'tl'],
      offset: popupOffset,
      ...defaultAlign,
    },
    rightTop: {
      points: ['tl', 'tr'],
      offset: popupOffset,
      ...defaultAlign,
    },
    left: {
      points: ['cr', 'cl'],
      offset: popupOffset,
      ...defaultAlign,
    },
    right: {
      points: ['cl', 'cr'],
      offset: popupOffset,
      ...defaultAlign,
    },
    top: {
      points: ['bc', 'tc'],
      offset: popupOffset,
      ...defaultAlign,
    },
    bottom: {
      points: ['tc', 'bc'],
      offset: popupOffset,
      ...defaultAlign,
    },
    topRight: {
      points: ['br', 'tr'],
      offset: popupOffset,
      ...defaultAlign,
    },
    bottomRight: {
      points: ['tr', 'br'],
      offset: popupOffset,
      ...defaultAlign,
    },
    rightBottom: {
      points: ['bl', 'br'],
      offset: popupOffset,
      ...defaultAlign,
    },
    leftBottom: {
      points: ['br', 'bl'],
      offset: popupOffset,
      ...defaultAlign,
    },
  };
}
