// @universe-design/react/src/menu/SubMenu.tsx
import type { FC } from 'react';
import React, { useMemo } from 'react';
import { RcSubMenu, useFullPath, MenuContext as RcMenuContext } from './internal-menu';
import clsx from 'clsx';
import stylex from '@universe-design/stylex';
import { cssTypographyBody0 } from '../stylex';
import { omit } from '../util';
import { isSubMenu } from './util';
import { MenuContext } from './MenuContext';
import { HorizontalMenuInk } from './HorizontalMenuInk';
import { getOffset } from './placements';
import { MenuTitle } from './MenuTitle';
import { EmptyElement } from '../shared/components/FirstLineAlign';
import './style';

export interface TitleEventEntity {
  /**
   * 菜单项的 key
   */
  key: string;
  /**
   * Event
   */
  domEvent: React.MouseEvent<HTMLElement> | React.KeyboardEvent<HTMLElement>;
}

export interface SubMenuProps {
  className?: string;
  /**
   * 是否禁用
   * @defaultValue false
   */
  disabled?: boolean;
  level?: number;
  title?: React.ReactNode;
  /**
   * 菜单图标
   */
  icon?: React.ReactNode;
  /**
   * style 透传
   */
  style?: React.CSSProperties;
  /**
   * 点击子菜单标题
   */
  onTitleClick?: (e: TitleEventEntity) => void;
  onTitleMouseEnter?: (e: TitleEventEntity) => void;
  onTitleMouseLeave?: (e: TitleEventEntity) => void;
  popupOffset?: [number, number];
  /**
   * 弹出层子菜单样式
   */
  popupClassName?: string;
  /**
   * 子菜单的菜单项
   */
  children?: React.ReactNode;
  textOverflow?: boolean;
  /**
   * 自定义次级菜单展开图标
   */
  expandIcon?: React.ReactNode;
}

const SubMenu: FC<SubMenuProps> = props => {
  const {
    icon,
    title,
    className,
    popupClassName,
    children,
    textOverflow: textOverflowInProps,
    disabled,
    popupOffset: popupOffsetProp,
  } = props;

  const context = React.useContext(MenuContext);
  /**
   * currentMode 是 当前子菜单的 mode,
   * 如 horizontal 的submenu mode 是 vertical
   */
  const { mode: currentMode } = React.useContext(RcMenuContext);
  //  此处 rootMode 是最上层用户传的 mode
  const { inlineCollapsed, textOverflow, prefixCls, mode: rootMode } = context;

  const parentPath = useFullPath();

  const subCls = clsx(
    className,
    `${prefixCls}__submenu--root-${rootMode}`,
    rootMode === 'normal' && stylex(cssTypographyBody0),
  );
  const popupOffset = getOffset(rootMode === 'normal' ? 'normal' : currentMode);

  const renderIcon = () =>
    icon && <div className={clsx(`${prefixCls}-icon-wrap`, disabled && `${prefixCls}-icon-wrap-disabled`)}>{icon}</div>;

  const renderSubMenuChildren = () => {
    if (!inlineCollapsed || parentPath.length) {
      return <MenuTitle showEllipsis={textOverflowInProps ?? textOverflow} title={title} />;
    }

    // 如果 collapsed 的情况
    // 1. 有 icon
    if (icon) {
      return null;
    }
    // 2. 没有 icon 收起模式下显示 icon 或是 children 的第一个字符
    if (typeof title === 'string') {
      return <div className={`${prefixCls}-inline-collapsed-noicon`}>{title.charAt(0)}</div>;
    }
  };

  const includeSubMenu = React.Children.toArray(children).find(isSubMenu);
  const cls = clsx(
    prefixCls,
    {
      [`${prefixCls}-submenu-popup-include-subMenu`]: includeSubMenu,
    },
    popupClassName,
    `${prefixCls}__submenu--root-${rootMode}`,
  );

  return (
    <MenuContext.Provider
      value={useMemo(
        () => ({
          ...context,
          firstLevel: false,
        }),
        [context],
      )}
    >
      <RcSubMenu
        {...omit(props, ['icon', 'textOverflow', 'className'])}
        popupOffset={popupOffsetProp ?? popupOffset}
        title={
          <>
            {renderIcon()}
            {renderSubMenuChildren()}
            <EmptyElement />
            <HorizontalMenuInk />
          </>
        }
        className={subCls}
        popupClassName={cls}
      />
    </MenuContext.Provider>
  );
};
if (__DEV__) {
  SubMenu.displayName = 'SubMenu';
}

export default SubMenu;
