// @universe-design/react/src/date-picker/panels/PanelHeaderSwitchers.tsx
import React from 'react';
import clsx from 'clsx';
import RightBoldOutlined from '@universe-design/icons-react/RightBoldOutlined';
import DownRightBoldOutlined from '@universe-design/icons-react/DownRightBoldOutlined';
import LeftBoldOutlined from '@universe-design/icons-react/LeftBoldOutlined';
import UpLeftBoldOutlined from '@universe-design/icons-react/UpLeftBoldOutlined';

import type { ButtonRefType, ButtonProps } from '../../button';
import { Button } from '../../button';
import { useCssScope } from '../../config-provider';

interface PickerHeaderSwitcherProps extends ButtonProps {
  /**
   * 是否作为占位符使用，可以在不展示展开按钮时保证父子元素对齐
   * @defaultValue false
   */
  asPlaceholder?: boolean;
}

export const usePickerHeaderSwitcherProps = ({
  className,
  asPlaceholder = false,
  ...restProps
}: PickerHeaderSwitcherProps): ButtonProps => {
  const prefixCls = useCssScope('picker')();
  const headerIconCls = `${prefixCls}-panel-header-icon`;

  return {
    ...restProps,
    type: 'icon',
    size: 'extra-small',
    className: clsx(headerIconCls, asPlaceholder && `${headerIconCls}__hidden`, className),
  };
};

const NextMonthSwitcher = React.forwardRef<ButtonRefType, PickerHeaderSwitcherProps>((props, ref) => {
  const buttonProps = usePickerHeaderSwitcherProps(props);

  return <Button ref={ref} {...buttonProps} icon={<RightBoldOutlined />} data-testid="ud__picker-nextMonth" />;
});

const NextYearSwitcher = React.forwardRef<ButtonRefType, PickerHeaderSwitcherProps>((props, ref) => {
  const buttonProps = usePickerHeaderSwitcherProps(props);

  return <Button ref={ref} {...buttonProps} icon={<DownRightBoldOutlined />} data-testid="ud__picker-nextYear" />;
});

const Next20YearsSwitcher = React.forwardRef<ButtonRefType, PickerHeaderSwitcherProps>((props, ref) => {
  const buttonProps = usePickerHeaderSwitcherProps(props);

  return <Button ref={ref} {...buttonProps} icon={<DownRightBoldOutlined />} data-testid="ud__picker-next20Years" />;
});

const PreMonthSwitcher = React.forwardRef<ButtonRefType, PickerHeaderSwitcherProps>((props, ref) => {
  const buttonProps = usePickerHeaderSwitcherProps(props);

  return <Button ref={ref} {...buttonProps} icon={<LeftBoldOutlined />} data-testid="ud__picker-prevMonth" />;
});

const PreYearSwitcher = React.forwardRef<ButtonRefType, PickerHeaderSwitcherProps>((props, ref) => {
  const buttonProps = usePickerHeaderSwitcherProps(props);

  return <Button ref={ref} {...buttonProps} icon={<UpLeftBoldOutlined />} data-testid="ud__picker-prevYear" />;
});

const Pre20YearsSwitcher = React.forwardRef<ButtonRefType, PickerHeaderSwitcherProps>((props, ref) => {
  const buttonProps = usePickerHeaderSwitcherProps(props);

  return <Button ref={ref} {...buttonProps} icon={<UpLeftBoldOutlined />} data-testid="ud__picker-prev20Years" />;
});

export {
  NextMonthSwitcher,
  NextYearSwitcher,
  Next20YearsSwitcher,
  PreMonthSwitcher,
  PreYearSwitcher,
  Pre20YearsSwitcher,
};
