// @universe-design/react/src/pin-input/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Basic from '../demo/0.basic';
import C1Size from '../demo/1.size';
import C2WithSeq from '../demo/2.with-seq';
import C3Status from '../demo/3.status';
import C4Mask from '../demo/4.mask';
import C5Alphanumeric from '../demo/5.alphanumeric';
if (!(C0Basic as Story).storyName) {
  (C0Basic as Story).storyName = '0. 基础用法';
  (C0Basic as Story).parameters = {
    docs: {
      description: {
        story: '简单的验证码输入。\n\n',
      },
    },
  };
}
if (!(C1Size as Story).storyName) {
  (C1Size as Story).storyName = '1. 不同尺寸';
  (C1Size as Story).parameters = {
    docs: {
      description: {
        story: '验证码输入组件默认内置三种预设尺寸。\n\n',
      },
    },
  };
}
if (!(C2WithSeq as Story).storyName) {
  (C2WithSeq as Story).storyName = '2. 带连字符';
  (C2WithSeq as Story).parameters = {
    docs: {
      description: {
        story: '带连字符的验证码输入，一共有三种符合规范的布局「定宽」、「可伸缩连字符间距」、「可伸缩输入框」。\n\n',
      },
    },
  };
}
if (!(C3Status as Story).storyName) {
  (C3Status as Story).storyName = '3. 其他状态';
  (C3Status as Story).parameters = {
    docs: {
      description: {
        story: '单字符输入拥有错误态和禁用态。\n\n',
      },
    },
  };
}
if (!(C4Mask as Story).storyName) {
  (C4Mask as Story).storyName = '4. 遮盖输入值';
  (C4Mask as Story).parameters = {
    docs: {
      description: {
        story:
          '当需要输入私密敏感信息时，需要遮盖输入值，与 `<input type="password"/>` 类型。可以使用 `mask` 属性做到。\n\n',
      },
    },
  };
}
if (!(C5Alphanumeric as Story).storyName) {
  (C5Alphanumeric as Story).storyName = '4. 输入数字与字母';
  (C5Alphanumeric as Story).parameters = {
    docs: {
      description: {
        story:
          '当需要允许输入数字和字母时，可以设置 type 为 `alphanumeric`。同理，也可以可以设置 type 为 `alphabetic`限制用户仅输入字母。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/PinInput/官网示例',
  tags: ['autodocs'],
};

export { C0Basic, C1Size, C2WithSeq, C3Status, C4Mask, C5Alphanumeric };
