// @byted-service/metrics/lib/auth.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Auth = void 0;
const util_1 = require("./util");
/**
 * _from: 1.4.0_
 *
 * **注意:** 2019.8.12开始，没有申请metrics query quota的脚本查询将被限制。
 * [相关文档](https://bytedance.feishu.cn/space/doc/doccnBm1rl8lsFqjtUnDzufVOyf)
 *
 * 申请后可以通过`Auth`实例获取token：
 * ```js
 * const { Auth } = require('@byted-service/metrics');
 *
 * const auth = new Auth({
 *   region: 'cn',
 *   app_name: 'YOUR_APP_NAME',
 *   app_secret: 'YOUR_APP_SECERT',
 * });
 *
 * const token = await auth.getToken()
 * ```
 */
class Auth {
    constructor(config) {
        /* istanbul ignore next */
        this.getCurrentToken = () => {
            return this.token;
        };
        /**
         * 更新token。如果token没有过期的话，metrics服务器返回的token可能不会有变化。
         */
        this.updateToken = async () => {
            const { authConfig, prefix } = this;
            const { app_name, app_secret, region } = authConfig;
            const url = `${prefix}?_region=${region}`;
            const retStr = await (0, util_1.post)(url, {
                app_name,
                app_secret
            }, {});
            const ret = JSON.parse(retStr);
            /* istanbul ignore next */
            if (ret.code !== 0) {
                throw new Error(`request failed: ${JSON.stringify(ret)}`);
            }
            const { access_token: token, expire } = ret;
            this.token = token;
            this.expireTime = expire * 1000;
        };
        /**
         * 获取token。如果没有token或token过期，则会带上数据去请求token；否则会直接返回本地的token。
         */
        this.getToken = async () => {
            const { token, expireTime } = this;
            if (!token || Date.now() >= expireTime) {
                await this.updateToken();
                /* istanbul ignore next */
                return this.token;
            }
            return this.token;
        };
        this.getExpireTime = () => {
            return this.expireTime;
        };
        if (!config ||
            !config.app_name ||
            !config.app_secret ||
            !config.region) {
            throw new Error('invalid Metrics Auth config: ' + JSON.stringify(config));
        }
        this.token = null;
        this.expireTime = 0;
        this.authConfig = Object.assign({}, config);
        this.prefix = config.prefix || (0, util_1.getAuthPrefix)(util_1.kHost);
    }
}
exports.Auth = Auth;
