// @byted-service/metrics/lib/consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.noop = exports.defaultOptions = exports.enableUnixDgram = exports.maxPacketSize = exports.MTU = exports.maxASCII = exports.maxTagLen = void 0;
const child_process_1 = require("child_process");
const env_1 = require("@byted-service/env");
exports.maxTagLen = 255;
exports.maxASCII = '\u007F';
const METRICS_MAX_PACKET_SIZE = Number(process.env.METRICS_MAX_PACKET_SIZE);
/**
 * 对于 IPv4 来说，推荐的 MTU 大小是 576
 * https://nodejs.org/docs/latest-v14.x/api/dgram.html#dgram_note_about_udp_datagram_size
 */
exports.MTU = 576;
/**
 * https://stackoverflow.com/questions/1098897/what-is-the-largest-safe-udp-packet-size-on-the-internet
 */
exports.maxPacketSize = METRICS_MAX_PACKET_SIZE > 0
    ? METRICS_MAX_PACKET_SIZE
    : process.platform === 'darwin'
        ? 9216
        : process.platform === 'linux' || process.platform === 'win32'
            ? 65507
            : exports.MTU;
exports.enableUnixDgram = process.env.ENABLE_METRICS_USE_UNIX_DGRAM === 'true';
/* istanbul ignore next */
const getGateway = () => {
    // gulu bootstrap.sh 里面定义了 ROUTE_IP
    if (process.env.ROUTE_IP) {
        return process.env.ROUTE_IP;
    }
    if ((0, env_1.isInTCE)()) {
        if (process.env.IS_HOST_NETWORK ||
            [
                'xflonline',
                'xflboe',
                'sinfonline',
                'sinfboe',
                'sinfonlinea',
                'sinfonlinec',
                'mya',
            ].includes((0, env_1.getIDC)())) {
            if ((0, env_1.isIPv6Only)()) {
                exports.maxPacketSize = 1024;
                return '::1';
            }
            else {
                return '127.0.0.1';
            }
        }
        // TCE 实例上，优先使用 MY_HOST_IP 获取，`ip route` 的方式不准确
        /**
         * 特殊 case，具体原因暂未查明
         * 1. 优先以 ipv4 栈发送数据
         * 2. 在 ipv6 only 下，修改最大包体积为 1k
         * 相关人员 @谢达 @赵杰裔
         */
        if (process.env.MY_HOST_IP) {
            return process.env.MY_HOST_IP;
        }
        if (process.env.MY_HOST_IPV6) {
            exports.maxPacketSize = 1024;
            return process.env.MY_HOST_IPV6;
        }
        const pattern = /default via (\S+)/;
        try {
            const route = ((0, child_process_1.execSync)('/opt/tiger/consul_deploy/bin/determine_dc.sh', {
                encoding: 'utf-8',
            }) || '').trim();
            const match = route.match(pattern);
            if (match) {
                return match[1];
            }
        }
        catch (err) {
            // IGNORE
        }
    }
    return (0, env_1.isIPv6Only)() ? '::1' : '127.0.0.1';
};
exports.defaultOptions = {
    debug: false,
    silentVerifying: (0, env_1.isInTCE)(),
    host: getGateway(),
    port: 9123,
    defaultPrefix: 'toutiao',
    bufferSize: 1000,
    flushInterval: 200,
    enablePreAggregate: false,
    logger: {
        fatal: console.log.bind(console),
        error: console.log.bind(console),
        warn: console.log.bind(console),
        info: console.log.bind(console),
        debug: console.log.bind(console),
        trace: console.log.bind(console),
    },
    disableBuiltinClusterTag: false,
};
const noop = () => {
    //
};
exports.noop = noop;
