// @byted-service/metrics/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Metrics = exports.getQueryPrefix = exports.getAuthPrefix = exports.kUsHost = exports.kHost = exports.Auth = exports.parseTimeString = exports.query = void 0;
const tslib_1 = require("tslib");
// NOTE export新的数据和方法应该直接挂在Metrics上，因为在这里export会被覆盖。
const metrics_1 = require("./metrics");
Object.defineProperty(exports, "Metrics", { enumerable: true, get: function () { return metrics_1.Metrics; } });
var query_1 = require("./query");
Object.defineProperty(exports, "query", { enumerable: true, get: function () { return query_1.query; } });
Object.defineProperty(exports, "parseTimeString", { enumerable: true, get: function () { return query_1.parseTimeString; } });
var auth_1 = require("./auth");
Object.defineProperty(exports, "Auth", { enumerable: true, get: function () { return auth_1.Auth; } });
var util_1 = require("./util");
Object.defineProperty(exports, "kHost", { enumerable: true, get: function () { return util_1.kHost; } });
Object.defineProperty(exports, "kUsHost", { enumerable: true, get: function () { return util_1.kUsHost; } });
Object.defineProperty(exports, "getAuthPrefix", { enumerable: true, get: function () { return util_1.getAuthPrefix; } });
Object.defineProperty(exports, "getQueryPrefix", { enumerable: true, get: function () { return util_1.getQueryPrefix; } });
tslib_1.__exportStar(require("./consts"), exports);
tslib_1.__exportStar(require("./typings"), exports);
exports.default = metrics_1.Metrics;
