// @byted-service/metrics/lib/pool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pool = exports.poolCache = exports.idCreator = void 0;
const net_1 = require("net");
const netpool_1 = require("@byted-service/netpool");
const idSym = Symbol.for('id');
exports.idCreator = (0, netpool_1.genIdCreator)('@byted-service/metrics');
exports.poolCache = {
    cache: new Map(),
    get(options) {
        let cacheItem = this.cache.get(options.logger);
        if (!cacheItem) {
            cacheItem = new Map();
        }
        let pool = cacheItem.get(options.host);
        if (!pool) {
            pool = new netpool_1.UDPPool({
                id: (0, exports.idCreator)(),
                connectOptions: {
                    type: (0, net_1.isIPv6)(options.host) ? 'udp6' : 'udp4',
                },
                logger: options.logger,
                create: (socket) => socket.unref(),
            });
            pool[idSym] = options;
        }
        cacheItem.set(options.host, pool);
        this.cache.set(options.logger, cacheItem);
        return pool;
    },
    values() {
        const pools = [];
        for (const [, cacheItem] of this.cache) {
            for (const [, pool] of cacheItem) {
                pools.push(pool);
            }
        }
        return pools;
    },
    delete(pool) {
        const { logger, host } = pool[idSym];
        const cacheItem = this.cache.get(logger);
        if (cacheItem) {
            cacheItem.delete(host);
            if (cacheItem.size === 0) {
                this.cache.delete(logger);
            }
        }
    },
};
exports.pool = new netpool_1.DynamicPool({
    poolCache: exports.poolCache,
});
