// @byted-service/metrics/lib/query.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.query = exports.parseTimeString = exports.getUnitMs = void 0;
const util_1 = require("./util");
const kDefaultMetricsOptions = {
    region: 'cn',
    timeout: 0
};
// private, for tests
function getUnitMs(unit) {
    switch (unit) {
        case 's': {
            return 1000;
        }
        case 'm': {
            return 60 * 1000;
        }
        case 'h': {
            return 60 * 60 * 1000;
        }
        case 'd': {
            return 24 * 60 * 60 * 1000;
        }
        case 'w': {
            return 7 * 24 * 60 * 60 * 1000;
        }
        case 'n': {
            // NOTE 假设一个月是30天
            return 30 * 24 * 60 * 60 * 1000;
        }
        case 'y': {
            // NOTE 假设一年是365天
            return 365 * 24 * 60 * 60 * 1000;
        }
        default:
            throw new Error(`invalid unit: ${unit}`);
    }
}
exports.getUnitMs = getUnitMs;
/**
 * 时间格式转换帮助方法
 *
 * opentsdb中可以指定时间格式`<amount><time unit>-ago`来方便地替代时间戳，如: `4h-ago`。目前公司的post风格的查询接口不支持用这样的时间格式来替代时间戳形式的`start`和`end`参数，需要使用此方法进行转换
 *
 * 相关文档可见：[Dates and Times](http://opentsdb.net/docs/build/html/user_guide/query/dates.html)
 * @param timeString 如: `4h-ago`
 */
function parseTimeString(timeString) {
    const res = /^(\d+)([smhdwny])-ago$/.exec(timeString);
    if (!res) {
        throw new Error(`invalid timeString: ${timeString}`);
    }
    const [, num, unit] = res;
    const dateNow = new Date();
    const value = dateNow.valueOf();
    const dateTs = value - Number(num) * getUnitMs(unit);
    return dateTs;
}
exports.parseTimeString = parseTimeString;
async function queryOnce(query, options = kDefaultMetricsOptions, forceUpdateToken = false) {
    const meta = Object.assign(Object.assign({}, kDefaultMetricsOptions), options);
    meta.prefix = meta.prefix || (0, util_1.getQueryPrefix)(util_1.kHost);
    const { region, timeout, prefix, auth } = meta;
    const url = `${prefix}?_region=${region}`;
    const requestOptions = {
        timeout,
        headers: {}
    };
    if (auth) {
        if (forceUpdateToken) {
            await auth.updateToken();
        }
        const token = await auth.getToken();
        requestOptions.headers = {
            Authorization: token
        };
    }
    return (0, util_1.safelyPost)(url, query, requestOptions);
}
/**
 * 数据查询，示例如下：
 *
 * ```js
 * const { Auth, query } = require('@byted-service/metrics');
 *
 * const auth = new Auth({
 *   region: 'cn',
 *   app_name: 'YOUR_APP_NAME',
 *   app_secret: 'YOUR_APP_SECERT',
 * });
 *
 * const options = {
 *   // NOTE 请确保region和Auth中的region一致，否则请求会失败
 *   region: 'cn',
 *   timeout: 0,
 *   // 如果没有传入auth实例，查询则不会带上token，查询可能会失败
 *   auth,
 * };
 *
 * (async () => {
 *   try {
 *     const data = await query({
 *       start: Date.now() - 5 * 60 * 1000,
 *       // 可在一个请求中聚合多个查询
 *       queries: [
 *         {
 *           metric: 'toutiao.byted-service.metrics.test.from.local',
 *           aggregator: 'avg',
 *         }
 *       ],
 *     }, options);
 *   } catch (err) {
 *     // 请求失败处理
 *   }
 * })();
 * ```
 *
 *
 * @param query 查询参数
 * @param options 请求参数设置
 *
 */
async function query(query, options = kDefaultMetricsOptions) {
    const queryTimes = 2;
    for (let i = 0; i < queryTimes; i += 1) {
        const ret = await queryOnce(query, options, i === 1);
        const { err, res, body } = ret;
        if (err) {
            throw err;
        }
        if (res.statusCode === 200) {
            return JSON.parse(body);
        }
        // body可能不是有效的json结构
        try {
            const msg = JSON.parse(body);
            if (i === 0 && String(msg.code) === '1' && res.statusCode === 401) {
                // 如果是token失效，则更新token，重新进行一次
                continue;
            }
        }
        catch (err) {
            // 这种情况下JSON.parse会抛出异常，我们代码分支直接走抛出异常的流程即可
        }
        throw new Error(`invalid status code: ${res.statusCode}, body: ${body}`);
    }
}
exports.query = query;
