// @byted-service/metrics/lib/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAgentPaths = exports.isAgentSidecar = exports.isValidString = exports.getTceTagsStr = exports.toKV = exports.post = exports.safelyPost = exports.getAuthPrefix = exports.getQueryPrefix = exports.kAuthPrefix = exports.kQueryPrefix = exports.kUsHost = exports.kHost = void 0;
const tslib_1 = require("tslib");
const http_1 = tslib_1.__importDefault(require("http"));
const url_1 = tslib_1.__importDefault(require("url"));
const env_1 = require("@byted-service/env");
const cn_json_1 = tslib_1.__importDefault(require("./config/cn.json"));
const us_json_1 = tslib_1.__importDefault(require("./config/us.json"));
const consts_1 = require("./consts");
const path_1 = tslib_1.__importDefault(require("path"));
exports.kHost = cn_json_1.default.kHost;
exports.kUsHost = us_json_1.default.kUsHost;
exports.kQueryPrefix = `${exports.kHost}/proxy/api/query`;
exports.kAuthPrefix = `${exports.kHost}/proxy/api/access_token`;
function getQueryPrefix(host) {
    return `${host}/proxy/api/query`;
}
exports.getQueryPrefix = getQueryPrefix;
function getAuthPrefix(host) {
    return `${host}/proxy/api/access_token`;
}
exports.getAuthPrefix = getAuthPrefix;
function safelyPost(url, query, options) {
    const body = JSON.stringify(query);
    // merge headers
    const headers = {
        'content-length': body.length,
    };
    if (typeof options.headers === 'object') {
        Object.assign(headers, options.headers);
        delete options.headers;
    }
    const params = Object.assign({
        method: 'post',
        headers,
    }, options);
    return new Promise((resolve, reject) => {
        const requestOptions = Object.assign(Object.assign({}, url_1.default.parse(url)), params);
        const req = http_1.default.request(requestOptions, (res) => {
            let response = Buffer.alloc(0);
            res.on('data', (data) => {
                response = Buffer.concat([response, data]);
            });
            res.on('end', () => {
                const body = response.toString('utf8');
                resolve({
                    res,
                    body,
                    err: null,
                });
            });
        });
        req.on('error', (err) => {
            resolve({
                res: null,
                body: '',
                err,
            });
        });
        req.end(body);
    });
}
exports.safelyPost = safelyPost;
async function post(url, query, options) {
    const { res, body, err } = await safelyPost(url, query, options);
    if (err) {
        throw err;
    }
    if (res.statusCode !== 200) {
        throw new Error(`invalid status code: ${res.statusCode}, body: ${body}`);
    }
    return body;
}
exports.post = post;
function toKV(tags, tagKeys, joiner = '|') {
    if (!tags) {
        return '';
    }
    let str = '';
    const keys = tagKeys || Object.keys(tags);
    if (keys.length <= 0) {
        return '';
    }
    for (let i = keys.length - 1; i > 0; i--) {
        const key = keys[i];
        const val = String(tags[key]);
        if (!(0, exports.isValidString)(key) || !(0, exports.isValidString)(val)) {
            continue;
        }
        str = `${joiner}${key}=${val}${str}`;
    }
    const key = keys[0];
    const val = String(tags[key]);
    if ((0, exports.isValidString)(key) && (0, exports.isValidString)(val)) {
        str = `${key}=${val}${str}`;
    }
    else {
        str = str.slice(1);
    }
    return str;
}
exports.toKV = toKV;
/**
 * 返回 tce 相关的 tags
 * https://code.byted.org/gopkg/metrics/blob/master/v3/option.go#L30
 * 目前 Golang 是需要用户手动的调用才会将 tce 相关的数据在打点时带上，
 * 但与 metrics 的同学沟通后，了解到后续 Golang 也会改为默认带上，所以 Node.js
 * 这边直接带上这几个 tag
 */
function getTceTagsStr(joiner = '|', disableBuiltinClusterTag = false) {
    const tags = {};
    if ((0, env_1.isInTCE)()) {
        tags.env_type = 'tce';
        tags.pod_name = (0, env_1.getPodName)();
    }
    tags._psm = (0, env_1.getPSM)();
    tags.deploy_stage = (0, env_1.getStage)();
    const ipv6 = (0, env_1.getIPv6)();
    if (ipv6) {
        tags.host_v6 = ipv6;
    }
    if ((0, env_1.getCluster)() && !disableBuiltinClusterTag) {
        tags.cluster = (0, env_1.getCluster)();
    }
    return toKV(tags);
}
exports.getTceTagsStr = getTceTagsStr;
const isValidString = (s) => {
    if (s.length === 0 || s.length > consts_1.maxTagLen) {
        return false;
    }
    for (let i = 0; i < s.length; i++) {
        if (s[i] > consts_1.maxASCII) {
            return false;
        }
    }
    return true;
};
exports.isValidString = isValidString;
const isAgentSidecar = () => {
    return process.env.METRICSERVER2_SIDECAR === '1';
};
exports.isAgentSidecar = isAgentSidecar;
const getAgentPaths = () => {
    if (process.env.BYTED_CI === '1') {
        return [path_1.default.resolve(__dirname, '../benchmark/ci.sock')];
    }
    if ((0, exports.isAgentSidecar)()) {
        return ['/sidecar/sock/metric.sock'];
    }
    else {
        return ['/opt/tmp/sock/metric.sock', '/tmp/metric.sock'];
    }
};
exports.getAgentPaths = getAgentPaths;
