// @byted-service/bytedtrace/lib/configsClient/configurator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Configurator = void 0;
const tslib_1 = require("tslib");
const confProvider = tslib_1.__importStar(require("../confProvider"));
const events_1 = require("events");
const logger_1 = require("../logger");
class Configurator {
    constructor() {
        this.functions = {};
    }
    static create(serviceName, cluster, framework, logger) {
        const that = new Configurator();
        that.logger = logger;
        that.event = new events_1.EventEmitter();
        confProvider.TCCClient.init(serviceName, cluster, framework);
        that.globalConfiger =
            new confProvider.Configer(confProvider.NAMESPACE_TRACER, that.event, logger);
        that.cbConfiger = new confProvider.Configer(confProvider.NAMESPACE_CB, that.event, logger);
        that.samplerConfiger =
            new confProvider.Configer(confProvider.NAMESPACE_SAMPLER, that.event, logger);
        that.event.on(confProvider.EVENT_CONFIG_CHANGE, (event) => {
            const f = that.functions[event.namespace];
            if (typeof f === 'function') {
                try {
                    f(event.config);
                    logger_1.debugLogger.debug('nameSpace[%s] synConfigs success.', event.namespace);
                }
                catch (error) {
                    /* istanbul ignore next */
                    that.logger.error('nameSpace[%s] synConfig failed: %o,config: %o,error: %o', event.namespace, event, event.config, error);
                }
            }
        });
        return that;
    }
    async syncInit() {
        await this.globalConfiger.syncSetDefaultConfig();
        await this.cbConfiger.syncSetDefaultConfig();
        await this.samplerConfiger.syncSetDefaultConfig();
    }
    asyncInit() {
        this.globalConfiger.asyncSetDefaultConfig();
        this.cbConfiger.asyncSetDefaultConfig();
        this.samplerConfiger.asyncSetDefaultConfig();
    }
    async pullGlobalTracerConfig(force) {
        return await this.globalConfiger.get(force);
    }
    async pullBreakerConfig(force) {
        return await this.cbConfiger.get(force);
    }
    async pullSamplerConfig(force) {
        return await this.samplerConfiger.get(force);
    }
    /**
     * 如果有变化需要触发注册的函数
     * The registered functions should be invoked if config has changed
     * @param key
     * @param f
     */
    register(key, f) {
        this.functions[key] = f;
    }
    close() {
        //
    }
}
exports.Configurator = Configurator;
