// @universe-design/react/lib/space/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Space = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _clsx = _interopRequireDefault(require("clsx"));
var _react = _interopRequireWildcard(require("react"));
var _hooks = require("../config-provider/hooks");
var _util = require("../util");
require("./style");
/**
 * 间距尺寸
 * @public
 */

const spaceSize = {
  small: 8,
  medium: 16,
  large: 24
};
function getNumberSize(size) {
  return typeof size === 'string' ? spaceSize[size] : size || 0;
}
/**
 * @public
 */

const Space = exports.Space = /*#__PURE__*/(0, _react.forwardRef)(function (_ref, ref) {
  let {
    children,
    align,
    size = 'small',
    direction = 'horizontal',
    className,
    wrap = false,
    split,
    style,
    keepEmpty = false,
    inline = true,
    shouldItemConfinedToContainer = false,
    itemKey,
    shouldPlaceItemInFlowRoot = true,
    ...restProps
  } = _ref;
  const cssScope = (0, _hooks.useCssScope)('space');
  const itemCssScope = (0, _hooks.useCssScope)('space__item');
  const childNodes = (0, _util.toFlatArray)(children, {
    keepEmpty
  });
  const [horizontalSize, verticalSize] = _react.default.useMemo(() => (Array.isArray(size) ? size : [size, size]).map(getNumberSize), [size]);
  const latestIndex = childNodes.length - 1;
  if (childNodes.length === 0) {
    return null;
  }
  const mergedAlign = (() => {
    if (align) {
      return align;
    }
    if (direction === 'horizontal') {
      return 'center';
    }
    if (!shouldPlaceItemInFlowRoot) {
      return 'start';
    }
  })();
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({}, restProps, {
    ref: ref,
    className: (0, _clsx.default)(cssScope(), cssScope(`-${direction}`), inline && cssScope('-inline'), mergedAlign && cssScope(`-align-${mergedAlign}`), className),
    style: {
      ...(wrap && {
        flexWrap: 'wrap'
      }),
      columnGap: horizontalSize / (split ? 2 : 1),
      rowGap: verticalSize,
      ...style
    }
  }), childNodes.map((node, index) => /*#__PURE__*/_react.default.createElement(_react.default.Fragment, {
    key: itemKey ? itemKey(index) : index
  }, shouldPlaceItemInFlowRoot ? /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(itemCssScope(), shouldItemConfinedToContainer && itemCssScope('-confined'))
  }, node) : node, index < latestIndex && split && /*#__PURE__*/_react.default.createElement("span", {
    className: itemCssScope('__split')
  }, split))));
});
