// @universe-design/react/src/date-picker/panels/DateCalendar/DateCalendarUI.tsx
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx } from '@universe-design/stylex/react';
import stylex from '@universe-design/stylex';
import clsx from 'clsx';
import isFunction from 'lodash/isFunction';
import isNil from 'lodash/isNil';
import min from 'lodash/min';
import noop from 'lodash/noop';
import type { Moment } from 'moment';
import moment from 'moment';
import React, { useEffect } from 'react';
import { useControlledState } from '@universe-design/react-utilities';
import { useCssScope } from '../../../config-provider';
import { Space } from '../../../space';
import { forwardHTMLRef } from '../../../util/forwardRef';
import type { MODE, OnSelect, StartOfWeek } from '../../interface';
import { sxPickerHeaderSwitchWrapper } from '../../sx';
import {
  Next20YearsSwitcher,
  NextMonthSwitcher,
  NextYearSwitcher,
  Pre20YearsSwitcher,
  PreMonthSwitcher,
  PreYearSwitcher,
} from '../PanelHeaderSwitchers';
import { getHigherLevelMode, getNext, getPrev } from '../base/utils';
import Body from '../picker/Body';
import { MonthYearSelector } from '../selector/MonthYearSelector';
import { CalendarUIProvider } from './DateCalendarUIContext';
import type { DateCalendarUIProps } from './interface';

import '../../style';
import { shouldCellDisabled as shouldCellDisabledFn } from '../../utils/disabled';

export const isDateValueEqual = (a: moment.Moment | undefined | null, b: moment.Moment | undefined | null) => {
  if (isNil(a) && isNil(b)) {
    return true;
  }
  if (a === b || a?.isSame(b)) {
    return true;
  }
  return false;
};

const defaultPreviewDate = () => false;
const defaultDisabledDate = () => false;

const DateCalendarUI = forwardHTMLRef<DateCalendarUIProps, 'div'>((props, ref) => {
  const {
    className,
    style,
    mode = 'date',
    pickerValue: pickerValueProp,
    hiddenCollapse,
    // monthYearSelectorProps
    onCollapsed,
    onMonthClick,
    onYearClick,

    // bodyProps
    onPickerValueChange,
    hiddenNext,
    hiddenPrev,
    headerRenderer,
    // context start
    defaultPickerValue, // pickerValue,
    selectedDates = [],
    selectedRange,
    onDateMouseEnter = noop,
    onDateMouseLeave = noop,
    onSelect = noop,
    previewDate = defaultPreviewDate,
    disabledDate = defaultDisabledDate,
    minDate,
    maxDate,
    startOfWeek = moment.localeData().firstDayOfWeek() as StartOfWeek,
    noOutOfViewWeek = false,
    cellRenderer,
    onContentMouseEnter = noop,
    onContentMouseLeave = noop,
    sx,
    janX,
    hideWeekNumbers = true,
    // context end
    ...restBodyProps
  } = props;
  const prefixCls = useCssScope('picker')();
  const datePanelPrefixCls = `${prefixCls}-date-panel`;

  const [localMode, setLocalMode] = React.useState<MODE>(mode);

  useEffect(() => {
    if (!janX) {
      return;
    }
    const currentLocale = moment.locale();
    moment.updateLocale(currentLocale, {
      week: {
        dow: startOfWeek,
        doy: 7 + startOfWeek - janX,
      },
    });
  }, [startOfWeek, janX]);

  // 当前面板切换到哪个时间点，并不是已选中的值或未提交的值
  const firstSelected = min([...selectedDates, ...(selectedRange || [])]);

  const [pickerValue = moment(), setViewDateState] = useControlledState<moment.Moment | undefined>(
    firstSelected ?? defaultPickerValue,
    pickerValueProp,
    {
      isEqual: isDateValueEqual,
    },
  );

  const setViewDate = (date: Moment) => {
    onPickerValueChange?.(date, { mode: localMode });
    setViewDateState(date);
  };

  const shouldCellDisabled = React.useCallback(
    (date: moment.Moment, _mode: MODE) =>
      shouldCellDisabledFn(date, {
        minDate,
        maxDate,
        disabledDate,
        mode: _mode,
      }),
    [disabledDate, maxDate, minDate],
  );

  const handleSelect: OnSelect = (newDate: moment.Moment) => {
    setViewDate(newDate);

    if (mode !== localMode) {
      setLocalMode(mode);
      return;
    }
    onSelect?.(newDate);
  };

  const handleCollapsed = (collapsed: boolean) => {
    if (isFunction(onCollapsed)) {
      onCollapsed(collapsed);
    }
    if (collapsed) {
      setLocalMode(mode);
    } else {
      const nextMode = getHigherLevelMode(mode);
      if (nextMode) {
        setLocalMode(nextMode);
      }
    }
  };

  const handleYearClick = () => {
    if (isFunction(onYearClick)) {
      onYearClick();
    }
    setLocalMode('year');
  };

  const handleMonthClick = () => {
    if (isFunction(onMonthClick)) {
      onMonthClick();
    }
    setLocalMode('month');
  };

  const goPrev = () => setViewDate(getPrev(pickerValue, localMode));
  const goNext = () => setViewDate(getNext(pickerValue, localMode));
  let PrevSwitcher, NextSwitcher;
  switch (localMode) {
    case 'date':
      PrevSwitcher = PreMonthSwitcher;
      NextSwitcher = NextMonthSwitcher;
      break;
    case 'month':
      PrevSwitcher = PreYearSwitcher;
      NextSwitcher = NextYearSwitcher;
      break;
    default:
      PrevSwitcher = Pre20YearsSwitcher;
      NextSwitcher = Next20YearsSwitcher;
      break;
  }

  const selectorCls = `${prefixCls}-panel-header-content__left`;

  // #region 自定义 Header 渲染
  const monthYearSelectorNode = (
    <MonthYearSelector
      mode={localMode}
      pickerValue={pickerValue}
      className={selectorCls}
      hiddenCollapse={hiddenCollapse || localMode === 'year'}
      collapsed={localMode === mode}
      onCollapsed={handleCollapsed}
      onYearClick={handleYearClick}
      onMonthClick={handleMonthClick}
    />
  );
  const prevBtnNode = <PrevSwitcher onClick={goPrev} />;
  const nextBtnNode = <NextSwitcher onClick={goNext} />;
  const defaultNode = (
    <div className={prefixCls + '-panel-header'}>
      <div className={`${prefixCls}-panel-header-inner`}>
        {monthYearSelectorNode}
        <Space size={12} className={stylex(sxPickerHeaderSwitchWrapper)}>
          {!hiddenPrev && prevBtnNode}
          {!hiddenNext && nextBtnNode}
        </Space>
      </div>
    </div>
  );
  const pickerHeaderNode = isFunction(headerRenderer)
    ? headerRenderer({
        prevBtnNode,
        nextBtnNode,
        defaultNode,
      })
    : defaultNode;
  // #endregion

  const calendarUIContext = {
    // defaultPickerValue,
    onPickerValueChange,
    selectedDates,
    selectedRange,
    onDateMouseEnter,
    onDateMouseLeave,
    onSelect: handleSelect,
    previewDate,
    shouldCellDisabled,
    pickerValue,
    startOfWeek,
    noOutOfViewWeek,
    cellRenderer,
    onContentMouseEnter,
    onContentMouseLeave,
    janX,
    hideWeekNumbers,
  };
  const datePanelPrefixClsWithMode = `${prefixCls}-date-panel__mode_${localMode}`;

  return (
    <CalendarUIProvider value={calendarUIContext}>
      <div
        className={clsx(
          className,
          datePanelPrefixCls,
          datePanelPrefixClsWithMode,
          !hideWeekNumbers && localMode === 'date' && `${datePanelPrefixCls}__withWeekNumbers`,
        )}
        style={style}
        ref={ref}
        sx={sx}
      >
        {pickerHeaderNode}
        <Body {...restBodyProps} mode={localMode} />
      </div>
    </CalendarUIProvider>
  );
});

export { DateCalendarUI };
