// @universe-design/react/es/table/hooks/useFilter/FilterMenu.js
import React from 'react';
import clsx from 'clsx';
import { Menu } from '../../../menu';
import { Button } from '../../../button';
import { Checkbox } from '../../../checkbox';
import { Radio } from '../../../radio';
import { Empty } from '../../../empty';
import { serializeFilterKey } from '../../utils';
import { ListDivider, ListItemLike } from '../../../shared/components/list';
function hasSubMenu(filters) {
  return filters.some(_ref => {
    let {
      children
    } = _ref;
    return children;
  });
}
export default function FilterMenu(props) {
  const {
    prefixCls,
    dropdownPrefixCls,
    tableLocale,
    filters,
    filterMultiple,
    openKeys,
    selectedKeys,
    onMenuClick,
    onSelectKeys,
    onOpenChange,
    onReset,
    onConfirm,
    getPopupContainer
  } = props;
  const renderFilterItems = colFilters => {
    if (colFilters.length === 0) {
      return /*#__PURE__*/React.createElement("div", {
        style: {
          margin: '12px 0'
        }
      }, /*#__PURE__*/React.createElement(Empty, {
        description: tableLocale.filterEmptyText,
        imageStyle: {
          height: 36
        }
      }));
    }
    return colFilters.map((filter, index) => {
      const key = serializeFilterKey(filter.value);
      if (filter.children) {
        return /*#__PURE__*/React.createElement(Menu.SubMenu, {
          key: key || index,
          title: filter.text,
          popupClassName: `${prefixCls}-dropdown-submenu`
        }, renderFilterItems(filter.children));
      }
      const Component = filterMultiple ? Checkbox : Radio;
      return /*#__PURE__*/React.createElement(Menu.Item, {
        key: filter.value !== undefined ? key : index
      }, /*#__PURE__*/React.createElement(Component, {
        checked: selectedKeys.includes(key)
      }), /*#__PURE__*/React.createElement("span", {
        className: `${prefixCls}-dropdown-label`
      }, filter.text));
    });
  };
  const triggerSelect = pramas => {
    var _pramas$selectedKeys;
    onSelectKeys({
      selectedKeys: (_pramas$selectedKeys = pramas.selectedKeys) !== null && _pramas$selectedKeys !== void 0 ? _pramas$selectedKeys : []
    });
  };
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Menu, {
    multiple: filterMultiple
    // prefixCls={`${dropdownPrefixCls}__menu`}
    ,
    className: clsx({
      [`${dropdownPrefixCls}__menu-without-submenu`]: !hasSubMenu(filters)
    }),
    mode: "normal",
    onClick: onMenuClick,
    onSelect: triggerSelect,
    onDeselect: triggerSelect,
    selectedKeys: selectedKeys,
    getPopupContainer: getPopupContainer,
    openKeys: openKeys,
    onOpenChange: onOpenChange,
    selectable: true,
    showCheckmark: false
  }, renderFilterItems(filters)), /*#__PURE__*/React.createElement(ListDivider, {
    fluid: true
  }), /*#__PURE__*/React.createElement(ListItemLike, {
    className: `${prefixCls}-dropdown-btns`
  }, /*#__PURE__*/React.createElement(Button, {
    type: "link",
    size: "small",
    disabled: selectedKeys.length === 0,
    onClick: onReset
  }, tableLocale.filterReset), /*#__PURE__*/React.createElement(Button, {
    type: "primary",
    size: "small",
    onClick: onConfirm
  }, tableLocale.filterConfirm)));
}
