// @universe-design/react/es/table/hooks/useFilter/FilterDropdown.js
import React, { useRef, useMemo, useCallback, useState, useEffect } from 'react';
import clsx from 'clsx';
import isEqual from 'lodash/isEqual';
import FilterOutlined from '@universe-design/icons-react/FilterOutlined';
import { Dropdown } from '../../../dropdown';
import { useForceUpdate } from '@universe-design/react-utilities';
import { isFunction } from '../../../util';
import FilterMenu from './FilterMenu';
import { IconButton } from '../../../button';
import { serializeFilterKeys, unSerializeFilterKeys } from '../../utils';
import { ListBox } from '../../../shared/components/list/List';
function FilterDropdown(props) {
  var _filterState$filtered;
  const {
    column,
    prefixCls,
    dropdownPrefixCls,
    tableLocale,
    columnKey,
    filterState,
    triggerFilter,
    children,
    filterMultiple,
    getPopupContainer
  } = props;
  const {
    filterDropdownVisible,
    onFilterDropdownVisibleChange
  } = column;
  const [visible, setVisible] = React.useState(false);
  const clickMenuRef = useRef(false);
  const forceUpdate = useForceUpdate();
  const filtered = Boolean(filterState && (((_filterState$filtered = filterState.filteredKeys) === null || _filterState$filtered === void 0 ? void 0 : _filterState$filtered.length) || filterState.forceFiltered));

  // dropdown visible
  const mergedVisible = typeof filterDropdownVisible === 'boolean' ? filterDropdownVisible : visible;
  const triggerVisible = newVisible => {
    setVisible(newVisible);
    if (onFilterDropdownVisibleChange) {
      onFilterDropdownVisibleChange(newVisible);
    }
  };

  // Select Keys
  const propFilteredKeys = useMemo(() => (filterState === null || filterState === void 0 ? void 0 : filterState.filteredKeys) || [], [filterState]);
  const filteredKeysSync = useRef(propFilteredKeys);
  const getFilteredKeysSync = () => filteredKeysSync.current;
  const setFilteredKeysSync = useCallback(value => {
    filteredKeysSync.current = value;
    forceUpdate();
  }, [filteredKeysSync, forceUpdate]);
  const onSelectKeys = useCallback(param => {
    setFilteredKeysSync(unSerializeFilterKeys(param.selectedKeys));
  }, [setFilteredKeysSync]);
  useEffect(() => {
    onSelectKeys({
      selectedKeys: serializeFilterKeys(propFilteredKeys) || []
    });
  }, [propFilteredKeys, onSelectKeys]);

  // Open Keys
  const [openKeys, setOpenKeys] = useState([]);

  // Open
  const onOpenChange = keys => {
    setOpenKeys(keys);
  };
  const onMenuClick = () => {
    clickMenuRef.current = true;
  };

  // Submit
  const internalTriggerFilter = keys => {
    triggerVisible(false);
    const mergedKeys = keys && keys.length ? keys : null;
    if (mergedKeys === null && (!filterState || !filterState.filteredKeys)) {
      return null;
    }
    if (isEqual(mergedKeys, filterState === null || filterState === void 0 ? void 0 : filterState.filteredKeys)) {
      return null;
    }
    triggerFilter({
      column,
      key: columnKey,
      filteredKeys: mergedKeys
    });
  };
  const onConfirm = () => {
    internalTriggerFilter(getFilteredKeysSync());
  };
  const onReset = () => {
    setFilteredKeysSync([]);
    internalTriggerFilter([]);
  };
  let dropdownContent, filterIconComp;
  if (isFunction(column.filterDropdown)) {
    const selectedKeys = getFilteredKeysSync() || [];
    dropdownContent = column.filterDropdown({
      prefixCls: `${dropdownPrefixCls}-custom`,
      setSelectedKeys: keys => onSelectKeys({
        selectedKeys: serializeFilterKeys(keys)
      }),
      selectedKeys,
      confirm: onConfirm,
      clearFilters: onReset,
      filters: column.filters,
      visible: mergedVisible
    });
  } else if (column.filterDropdown) {
    dropdownContent = column.filterDropdown;
  } else {
    const selectedKeys = serializeFilterKeys(getFilteredKeysSync() || []);
    dropdownContent = /*#__PURE__*/React.createElement(FilterMenu, {
      prefixCls: `${prefixCls}-filter`,
      tableLocale: tableLocale,
      dropdownPrefixCls: dropdownPrefixCls,
      filterMultiple: filterMultiple,
      filters: column.filters || [],
      openKeys: openKeys,
      selectedKeys: selectedKeys,
      onMenuClick: onMenuClick,
      onSelectKeys: onSelectKeys,
      onOpenChange: onOpenChange,
      onConfirm: onConfirm,
      onReset: onReset,
      getPopupContainer: getPopupContainer
    });
  }
  if (isFunction(column.filterIcon)) {
    filterIconComp = column.filterIcon(filtered);
  } else if (column.filterIcon) {
    filterIconComp = column.filterIcon;
  } else {
    filterIconComp = /*#__PURE__*/React.createElement(FilterOutlined, null);
  }
  const menu = /*#__PURE__*/React.createElement(ListBox, null, dropdownContent);
  const dropDownNode = /*#__PURE__*/React.createElement(Dropdown, {
    overlay: menu,
    trigger: ['click'],
    visible: mergedVisible,
    onVisibleChange: newVisible => {
      if (newVisible && propFilteredKeys !== undefined) {
        setFilteredKeysSync(propFilteredKeys);
      }
      if (!clickMenuRef.current) {
        triggerVisible(newVisible);
      }
      if (!clickMenuRef.current && !newVisible && !column.filterDropdown) {
        onConfirm();
      }
      clickMenuRef.current = false;
    },
    getPopupContainer: getPopupContainer,
    overlayClassName: `${prefixCls}-filter-dropdown`
  }, /*#__PURE__*/React.createElement(IconButton, {
    color: filtered ? 'primary' : undefined,
    selected: filtered || mergedVisible,
    icon: filterIconComp,
    size: "small"
  }));
  return /*#__PURE__*/React.createElement(React.Fragment, null, children, /*#__PURE__*/React.createElement("span", {
    className: clsx(`${prefixCls}-column-filter`, {
      [`${prefixCls}-filter-trigger-container-open`]: mergedVisible
    }),
    onClick: event => {
      event.stopPropagation();
    }
  }, dropDownNode));
}
export default FilterDropdown;
