// @universe-design/react/lib/color-picker/ColorPickerContext.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ColorPickerProvider = void 0;
exports.useColorPicker = useColorPicker;
exports.useColorPickerContextSelector = exports.useColorPickerContext = void 0;
var _react = _interopRequireWildcard(require("react"));
var _isEqual = _interopRequireDefault(require("lodash/isEqual"));
var _allLight = require("@universe-design/styles/cjs/all.light.atom");
var _reactUtilities = require("@universe-design/react-utilities");
var _util = require("./util");
var _hooks = require("../config-provider/hooks");
var _util2 = require("../util");
var _useSyncState = _interopRequireDefault(require("../hooks/useSyncState"));
var _colors = require("@universe-design/colors");
var _useContextSelector = require("use-context-selector");
var _const = require("./const");
const isColorEqual = (pre, cur) => {
  if (pre === undefined && cur === undefined) {
    return true;
  } else if (pre === undefined || cur === undefined) {
    return false;
  } else {
    return new _colors.TinyColor(pre).equals(cur);
  }
};
const {
  useContext: useColorPickerContext,
  useContextSelector: useColorPickerContextSelector,
  Provider: ColorPickerProvider
} = (0, _reactUtilities.createContextHook)('ColorPickerContext', {
  strict: true,
  customizedContextMethods: {
    createContext: _useContextSelector.createContext,
    useContext: _useContextSelector.useContext,
    useContextSelector: _useContextSelector.useContextSelector
  }
});
exports.ColorPickerProvider = ColorPickerProvider;
exports.useColorPickerContextSelector = useColorPickerContextSelector;
exports.useColorPickerContext = useColorPickerContext;
const defaultTransformDisplayColor = c => c.hex;
function useColorPicker(props) {
  const {
    defaultValue = _allLight.R500,
    value: valueProp,
    showDarkCheckmark = _util.defaultShowDarkCheckmark,
    prefixCls,
    onChange,
    transformDisplayColor: transformDisplayColorProp = defaultTransformDisplayColor
  } = props;
  const cssScope = (0, _hooks.useCssScope)('color-picker', prefixCls);
  const [controlledValue, _setControlledValue] = (0, _reactUtilities.useControlledState)(defaultValue, valueProp, {
    isEqual: _isEqual.default,
    isControlledOverride: 'value' in props
  });
  const setControlledValue = (0, _react.useCallback)((color, event) => {
    if ((0, _util2.isNonNullable)(color)) {
      const colorObj = (0, _util.formatColor)(color);
      (0, _util2.isFunction)(onChange) && onChange(colorObj.hex, colorObj, event);
      return;
    }
    (0, _util2.isFunction)(onChange) && onChange(undefined, undefined, event);
    _setControlledValue(color);
  }, [_setControlledValue, onChange]);
  const [syncValue, setSyncValue] = (0, _useSyncState.default)(controlledValue, {
    isEqual: isColorEqual
  });
  const setValue = _react.default.useCallback((_value, event) => {
    setSyncValue(_value);
    setControlledValue(_value, event);
  }, [setControlledValue, setSyncValue]);
  const formattedColor = _react.default.useMemo(() => syncValue !== undefined ? (0, _util.formatColor)(syncValue) : syncValue, [syncValue]);
  const transformDisplayColor = _react.default.useCallback(color => (0, _util.formatColor)(transformDisplayColorProp(color)), [transformDisplayColorProp]);
  const displayColor = _react.default.useMemo(() => formattedColor && transformDisplayColor(formattedColor), [formattedColor, transformDisplayColor]);
  return {
    value: syncValue,
    setValue,
    showDarkCheckmark,
    rootCssScope: cssScope,
    color: formattedColor,
    hsl: (formattedColor !== null && formattedColor !== void 0 ? formattedColor : _const.defaultColor).hsl,
    hsv: (formattedColor !== null && formattedColor !== void 0 ? formattedColor : _const.defaultColor).hsv,
    rgb: (formattedColor !== null && formattedColor !== void 0 ? formattedColor : _const.defaultColor).rgb,
    hex: (formattedColor !== null && formattedColor !== void 0 ? formattedColor : _const.defaultColor).hex,
    displayColor,
    transformDisplayColor
  };
}
