// @universe-design/react/lib/color-picker/type.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HrgbKey = exports.ColorPickerStep = exports.ColorPickerChangeSource = void 0;
/**
 * @public
 */
/**
 * @public
 */
let HrgbKey = exports.HrgbKey = /*#__PURE__*/function (HrgbKey) {
  HrgbKey["HEX"] = "hex";
  HrgbKey["R"] = "r";
  HrgbKey["G"] = "g";
  HrgbKey["B"] = "b";
  return HrgbKey;
}({});
let ColorPickerStep = exports.ColorPickerStep = /*#__PURE__*/function (ColorPickerStep) {
  ColorPickerStep[ColorPickerStep["inHome"] = 0] = "inHome";
  ColorPickerStep[ColorPickerStep["inMoreColor"] = 1] = "inMoreColor";
  return ColorPickerStep;
}({});
/**
 * @public
 * 变动来源，如手动在输入框输入、调节了色调滑块等
 */
let ColorPickerChangeSource = exports.ColorPickerChangeSource = /*#__PURE__*/function (ColorPickerChangeSource) {
  ColorPickerChangeSource["INPUT"] = "input";
  ColorPickerChangeSource["COLOR_BLOCK"] = "colorBlock";
  ColorPickerChangeSource["HUE"] = "hue";
  ColorPickerChangeSource["SATURATION"] = "saturation";
  return ColorPickerChangeSource;
}({});
/**
 * @public
 */
/**
 * `<PrimitiveColorPicker />` 和 `<PaletteColorPicker />` 组件的属性。只包含纯数据逻辑的 `ColorPicker` 组件，适用于自定义组合内容的场景。
 * @public
 */
/**
 * @public
 */
