// @universe-design/react/lib/menu/MenuTitle.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MenuTitle = void 0;
var _u1wwc0yi = _interopRequireDefault(require("./../.stylex/u1wwc0yi"));
var _u12n1ij = _interopRequireDefault(require("./../.stylex/u12n1ij7"));
var _u1g5jyy = _interopRequireDefault(require("./../.stylex/u1g5jyy"));
var _u99gbyh = _interopRequireDefault(require("./../.stylex/u99gbyh"));
var _urmav4p = _interopRequireDefault(require("./../.stylex/urmav4p"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _clsx = _interopRequireDefault(require("clsx"));
var _react = _interopRequireDefault(require("react"));
var _text = require("../text");
var _MenuContext = require("./MenuContext");
var _MenuContext2 = require("./internal-menu/context/MenuContext");
//.u1qumt79{line-height:inherit;}
//.uf5ikls{font-weight:inherit;}
//.ujoy568{font-size:inherit;}
//.uqerorx{flex-grow:1;}
//.u10tiqix{min-width:0px;}
const sxMenuTitleContent = {
  "props": {
    "Bg96gwp": _u1wwc0yi.default,
    "Bhrd7zp": _u12n1ij.default,
    "Be2twd7": _u1g5jyy.default,
    "Bh6795r": _u99gbyh.default,
    "Bf4jedk": _urmav4p.default
  }
};
const MenuTitle = _ref => {
  let {
    showEllipsis,
    title,
    style
  } = _ref;
  const {
    mode: rcMode
  } = _react.default.useContext(_MenuContext2.RcMenuContext);
  const {
    prefixCls
  } = _react.default.useContext(_MenuContext.MenuContext);
  if (showEllipsis) {
    const lineClamp = rcMode === 'vertical' || rcMode === 'inline' ? 2 : 1;
    return /*#__PURE__*/_react.default.createElement(_text.Text, {
      as: "div",
      className: (0, _clsx.default)(`${prefixCls}-item-title-content`, `${prefixCls}-item-title-content-text-overflow`),
      style: style,
      sx: [sxMenuTitleContent],
      lineClamp: lineClamp
    }, title);
  }
  return /*#__PURE__*/_react.default.createElement(_text.Text, {
    as: "div",
    style: style,
    className: `${prefixCls}-item-title-content`,
    sx: [sxMenuTitleContent]
  }, title);
};
exports.MenuTitle = MenuTitle;
