// @universe-design/react/lib/menu/Menu.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _RightBoldOutlined = _interopRequireDefault(require("@universe-design/icons-react/RightBoldOutlined"));
var _internalMenu = _interopRequireDefault(require("./internal-menu"));
var _layout = require("../layout");
var _hooks = require("../config-provider/hooks");
var _MenuContext = require("./MenuContext");
var _motion = require("../shared/motion");
var _util = require("./util");
var _util2 = require("../util");
var _DropdownMenuContext = require("../dropdown/DropdownMenuContext");
var _MenuMore = _interopRequireDefault(require("./MenuMore"));
var _placements = require("./placements");
var _constants = require("./constants");
require("./style");
/**
 * @docgen
 */

const Menu = props => {
  const {
    mode: modeProp,
    inlineCollapsed = false,
    className,
    prefixCls: prefixClsInProps,
    expandIcon,
    textOverflow = modeProp !== 'normal',
    getPopupContainer = () => document.body,
    subMenuCloseDelay,
    subMenuOpenDelay,
    inlineIndent = _constants.navMenuDefaults.inlineIndent,
    showCheckmark = true,
    onClick: onClickProp,
    selectable: selectableProp,
    builtinPlacements: builtinPlacementsProp,
    overflowedIndicator: overflowedIndicatorProp,
    selectedKeys,
    defaultSelectedKeys
  } = props;
  const rootClasPrefix = (0, _hooks.useClassPrefix)();
  const siderCtx = (0, _react.useContext)(_layout.SiderContext);
  const dropdownCtx = (0, _react.useContext)(_DropdownMenuContext.DropdownMenuContext);
  /**
   * 对于menu的mode属性，首先采用用户定义的mode，再考虑:
   * 如果是dropdown中的menu，即存在dropdownCtx，则mode定义为"normal"
   * 如果是普通的menu，则mode定义为"vertical"
   */
  const mergedMode = _react.default.useMemo(() =>
  // eslint-disable-next-line no-nested-ternary
  (0, _util2.isNonNullable)(modeProp) ? modeProp : dropdownCtx ? 'normal' : 'vertical', [dropdownCtx, modeProp]);
  const mergedPlacements = _react.default.useMemo(() => {
    if (builtinPlacementsProp) {
      return builtinPlacementsProp;
    }
    return (0, _placements.getPlacements)(mergedMode);
  }, [builtinPlacementsProp, mergedMode]);
  /**
   * 对于menu的selectable属性，首先采用用户定义的selectable，再考虑:
   * 如果传递了 selectedKeys, 也认为 selectable 为 true
   * 如果是dropdown中的menu，即存在dropdownCtx，则selectable定义为false
   * 如果是普通的menu，则selectable定义为true
   */
  const mergedSelectable = _react.default.useMemo(() => {
    if ((0, _util2.isNonNullable)(selectableProp)) {
      return selectableProp;
    }
    if ((0, _util2.isNonNullable)(selectedKeys) || (0, _util2.isNonNullable)(defaultSelectedKeys)) {
      return true;
    }
    return !dropdownCtx;
  }, [defaultSelectedKeys, dropdownCtx, selectableProp, selectedKeys]);
  const cssScope = (0, _hooks.useCssScope)(mergedMode === 'normal' ? 'menu-normal' : 'menu');
  const prefixCls = prefixClsInProps || cssScope();
  process.env.NODE_ENV !== "production" ? (0, _util2.warning)(
  // eslint-disable-next-line react/destructuring-assignment
  !(siderCtx && props.inlineCollapsed !== undefined), (0, _util2.devWarning)('Menu', 'prop `inlineCollapsed` will be ignored inside a <Sider /> component')) : void 0;
  const mergedInlineCollapsed = siderCtx ? siderCtx.siderCollapsed : inlineCollapsed;
  const {
    children
  } = props;
  const includeSubMenu = _react.default.Children.toArray(children).find(_util.isSubMenu);
  const menuClassName = (0, _clsx.default)(className, dropdownCtx === null || dropdownCtx === void 0 ? void 0 : dropdownCtx.className, {
    [`${prefixCls}-inline-collapsed`]: mergedInlineCollapsed,
    [`${prefixCls}-include-subMenu`]: includeSubMenu
  }, 'ud-scrollbar');
  const defaultMotions = {
    horizontal: {
      motionName: `${rootClasPrefix}-slide-up`
    },
    inline: (0, _motion.collapseMotion)(rootClasPrefix),
    other: {
      motionName: `${rootClasPrefix}-zoom-big`
    }
  };
  const rcMode = mergedMode === 'normal' ? 'vertical' : mergedMode;
  return /*#__PURE__*/_react.default.createElement(_MenuContext.MenuContext.Provider, {
    value: (0, _react.useMemo)(() => ({
      prefixCls,
      inlineCollapsed: mergedInlineCollapsed,
      mode: mergedMode,
      selectable: mergedSelectable,
      textOverflow,
      firstLevel: true,
      showCheckmark
    }), [mergedInlineCollapsed, mergedMode, mergedSelectable, prefixCls, showCheckmark, textOverflow])
  }, /*#__PURE__*/_react.default.createElement(_internalMenu.default, (0, _extends2.default)({
    builtinPlacements: mergedPlacements,
    getPopupContainer: getPopupContainer
  }, (0, _util2.omit)(props, ['showCheckmark', 'textOverflow']), {
    overflowedIndicator: overflowedIndicatorProp !== null && overflowedIndicatorProp !== void 0 ? overflowedIndicatorProp : /*#__PURE__*/_react.default.createElement(_MenuMore.default, null),
    onClick: _react.default.useCallback(e => {
      (0, _util2.isFunction)(onClickProp) && onClickProp(e);
      (0, _util2.isFunction)(dropdownCtx === null || dropdownCtx === void 0 ? void 0 : dropdownCtx.onClick) && (dropdownCtx === null || dropdownCtx === void 0 ? void 0 : dropdownCtx.onClick(e));
    }, [dropdownCtx, onClickProp]),
    mode: rcMode,
    selectable: mergedSelectable,
    inlineCollapsed: mergedInlineCollapsed,
    className: menuClassName,
    prefixCls: prefixCls,
    direction: "ltr",
    expandIcon: () => /*#__PURE__*/_react.default.createElement("span", {
      className: `${prefixCls}-submenu-arrow`
    }, expandIcon || /*#__PURE__*/_react.default.createElement(_RightBoldOutlined.default, {
      className: `${prefixCls}-submenu-arrow-icon`
    })),
    defaultMotions: defaultMotions
  }, (0, _util.milliSeconds2Seconds)({
    subMenuOpenDelay,
    subMenuCloseDelay
  }), {
    inlineIndent: inlineIndent
  })));
};
var _default = exports.default = Menu;
