// @universe-design/react/lib/menu/MenuItem.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _uevmt6d = _interopRequireDefault(require("./../.stylex/uevmt6d"));
var _uidkx = _interopRequireDefault(require("./../.stylex/uidkx17"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _ListCheckBoldOutlined = _interopRequireDefault(require("@universe-design/icons-react/ListCheckBoldOutlined"));
var _internalMenu = require("./internal-menu");
var _clsx = _interopRequireDefault(require("clsx"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _stylex2 = require("../stylex");
var _MenuContext = require("./MenuContext");
var _tooltip = require("../tooltip");
var _util = require("../util");
var _HorizontalMenuInk = require("./HorizontalMenuInk");
var _MenuTitle = require("./MenuTitle");
var _FirstLineAlign = require("../shared/components/FirstLineAlign");
require("./style");
//.u64fuq3{width:20px;}
//.ui64zpg{flex-shrink:0;}
/**
 * @docgen
 * @public
 */
const MenuItemSuffixIcon = _ref => {
  let {
    isSelected
  } = _ref;
  const {
    prefixCls,
    mode: rootMode,
    inlineCollapsed,
    showCheckmark,
    selectable
  } = React.useContext(_MenuContext.MenuContext);
  return selectable && showCheckmark && rootMode === 'normal' && !inlineCollapsed ? /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("span", {
    className: (0, _stylex.default)({
      "props": {
        "a9b677": _uevmt6d.default,
        "Bnnss6s": _uidkx.default
      }
    })
  }), isSelected && /*#__PURE__*/React.createElement(_ListCheckBoldOutlined.default, {
    className: `${prefixCls}-item-suffix-icon`
  })) : null;
};
const MenuItem = props => {
  const {
    className,
    children,
    title,
    icon,
    textOverflow: textOverflowInProps,
    disabled,
    ...rest
  } = props;
  const {
    prefixCls,
    inlineCollapsed,
    textOverflow,
    firstLevel,
    mode: rootMode
  } = React.useContext(_MenuContext.MenuContext);
  let tooltipTitle = title;
  if (typeof title === 'undefined') {
    tooltipTitle = firstLevel ? children : '';
  } else if (title === false) {
    tooltipTitle = '';
  }
  const tooltipProps = {
    title: tooltipTitle
  };
  if (!inlineCollapsed) {
    // tooltipProps.title = null; // 这个属性变化会导致 children remount，从而引起 children transition 动画失效
    // Reset `visible` to fix control mode tooltip display not correct
    tooltipProps.visible = false;
  }
  const childrenLength = (0, _util.toFlatArray)(children).length;
  const cls = (0, _clsx.default)(className, `${prefixCls}-item--root-${rootMode}`, {
    [`${prefixCls}-item-only-child`]: (icon ? childrenLength + 1 : childrenLength) === 1
  }, (0, _stylex.default)(rootMode === 'normal' && _stylex2.cssTypographyBody0));
  return /*#__PURE__*/React.createElement(_tooltip.Tooltip, (0, _extends2.default)({}, tooltipProps, {
    overlayClassName: `${prefixCls}-inline-collapsed-tooltip`,
    placement: "right"
  }), /*#__PURE__*/React.createElement(_internalMenu.RcMenuItem, (0, _extends2.default)({}, rest, {
    className: cls,
    title: typeof title === 'string' ? title : undefined,
    itemIcon: MenuItemSuffixIcon,
    disabled: disabled
  }), (() => icon && /*#__PURE__*/React.createElement("div", {
    className: (0, _clsx.default)(`${prefixCls}-icon-wrap`, disabled && `${prefixCls}-icon-wrap-disabled`)
  }, icon))(), (() => {
    const collapsedItem = inlineCollapsed && firstLevel;
    // 收起模式下显示 icon 或是 children 的第一个字符
    if (collapsedItem) {
      if (!icon && typeof children === 'string') {
        return /*#__PURE__*/React.createElement("div", {
          className: `${prefixCls}-inline-collapsed-noicon`
        }, children.charAt(0));
      }
    }
    const showEllipsis = !collapsedItem && (textOverflowInProps !== null && textOverflowInProps !== void 0 ? textOverflowInProps : textOverflow);
    return /*#__PURE__*/React.createElement(_MenuTitle.MenuTitle, {
      style: collapsedItem ? {
        width: 0,
        overflow: 'hidden'
      } : {},
      showEllipsis: showEllipsis,
      title: children
    });
  })(), /*#__PURE__*/React.createElement(_FirstLineAlign.EmptyElement, null), /*#__PURE__*/React.createElement(_HorizontalMenuInk.HorizontalMenuInk, null)));
};
if (process.env.NODE_ENV !== "production") {
  MenuItem.displayName = 'MenuItem';
}
var _default = exports.default = MenuItem;
