// @universe-design/react/lib/menu/SubMenu.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _internalMenu = require("./internal-menu");
var _clsx = _interopRequireDefault(require("clsx"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _stylex2 = require("../stylex");
var _util = require("../util");
var _util2 = require("./util");
var _MenuContext = require("./MenuContext");
var _HorizontalMenuInk = require("./HorizontalMenuInk");
var _placements = require("./placements");
var _MenuTitle = require("./MenuTitle");
var _FirstLineAlign = require("../shared/components/FirstLineAlign");
require("./style");
const SubMenu = props => {
  const {
    icon,
    title,
    className,
    popupClassName,
    children,
    textOverflow: textOverflowInProps,
    disabled,
    popupOffset: popupOffsetProp
  } = props;
  const context = _react.default.useContext(_MenuContext.MenuContext);
  /**
   * currentMode 是 当前子菜单的 mode,
   * 如 horizontal 的submenu mode 是 vertical
   */
  const {
    mode: currentMode
  } = _react.default.useContext(_internalMenu.MenuContext);
  //  此处 rootMode 是最上层用户传的 mode
  const {
    inlineCollapsed,
    textOverflow,
    prefixCls,
    mode: rootMode
  } = context;
  const parentPath = (0, _internalMenu.useFullPath)();
  const subCls = (0, _clsx.default)(className, `${prefixCls}__submenu--root-${rootMode}`, rootMode === 'normal' && (0, _stylex.default)(_stylex2.cssTypographyBody0));
  const popupOffset = (0, _placements.getOffset)(rootMode === 'normal' ? 'normal' : currentMode);
  const includeSubMenu = _react.default.Children.toArray(children).find(_util2.isSubMenu);
  const cls = (0, _clsx.default)(prefixCls, {
    [`${prefixCls}-submenu-popup-include-subMenu`]: includeSubMenu
  }, popupClassName, `${prefixCls}__submenu--root-${rootMode}`);
  return /*#__PURE__*/_react.default.createElement(_MenuContext.MenuContext.Provider, {
    value: (0, _react.useMemo)(() => ({
      ...context,
      firstLevel: false
    }), [context])
  }, /*#__PURE__*/_react.default.createElement(_internalMenu.RcSubMenu, (0, _extends2.default)({}, (0, _util.omit)(props, ['icon', 'textOverflow', 'className']), {
    popupOffset: popupOffsetProp !== null && popupOffsetProp !== void 0 ? popupOffsetProp : popupOffset,
    title: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, (() => icon && /*#__PURE__*/_react.default.createElement("div", {
      className: (0, _clsx.default)(`${prefixCls}-icon-wrap`, disabled && `${prefixCls}-icon-wrap-disabled`)
    }, icon))(), (() => {
      if (!inlineCollapsed || parentPath.length) {
        return /*#__PURE__*/_react.default.createElement(_MenuTitle.MenuTitle, {
          showEllipsis: textOverflowInProps !== null && textOverflowInProps !== void 0 ? textOverflowInProps : textOverflow,
          title: title
        });
      }

      // 如果 collapsed 的情况
      // 1. 有 icon
      if (icon) {
        return null;
      }
      // 2. 没有 icon 收起模式下显示 icon 或是 children 的第一个字符
      if (typeof title === 'string') {
        return /*#__PURE__*/_react.default.createElement("div", {
          className: `${prefixCls}-inline-collapsed-noicon`
        }, title.charAt(0));
      }
    })(), /*#__PURE__*/_react.default.createElement(_FirstLineAlign.EmptyElement, null), /*#__PURE__*/_react.default.createElement(_HorizontalMenuInk.HorizontalMenuInk, null)),
    className: subCls,
    popupClassName: cls
  })));
};
if (process.env.NODE_ENV !== "production") {
  SubMenu.displayName = 'SubMenu';
}
var _default = exports.default = SubMenu;
