// @universe-design/react/src/upload/Dragger.tsx
import type { HTMLAttributes } from 'react';
import React, { useState } from 'react';
import clsx from 'clsx';
import stylex from '@universe-design/stylex';
import FileAddColorful from '@universe-design/icons-react/FileAddColorful';
import FileAddDisableColorful from '@universe-design/icons-react/FileAddDisableColorful';
import { Upload } from './Upload';
import type { DragAreaConfig, DraggerUploadProps, UploadRef } from './interface';
import { useUploadContext } from './hooks/UploadContext';
import { useDragAreaSx } from './sx';
import { forwardHTMLRef } from '../util/forwardRef';

export const UploadDragArea = forwardHTMLRef<DragAreaConfig, 'div'>((props, ref) => {
  const {
    dragIcon: dragIconProp,
    dragDescription,
    dragTitle,
    className,
    onMouseEnter: onMouseEnterProp,
    onMouseLeave: onMouseLeaveProp,
    onDragLeave: onDragLeaveProp,
    onDragOver: onDragOverProp,
    ...resetProps
  } = props;
  const { rootPrefixCls, disabled } = useUploadContext();
  const [isActive, setIsActive] = useState(false);
  const {
    sxDisabledText,
    sxDragArea,
    sxDragAreaActive,
    sxDragAreaDescription,
    sxDragAreaDisabled,
    sxDragAreaIcon,
    sxDragAreaTitle,
  } = useDragAreaSx(rootPrefixCls);

  const classNames = clsx([
    className,
    stylex(sxDragArea, isActive && sxDragAreaActive, disabled && sxDragAreaDisabled),
  ]);

  const dragIcon =
    dragIconProp ?? ((disabledArg: boolean) => (disabledArg ? <FileAddDisableColorful /> : <FileAddColorful />));
  const icon = typeof dragIcon === 'function' ? dragIcon(disabled) : dragIcon;

  const events: HTMLAttributes<HTMLDivElement> = React.useMemo(() => {
    if (disabled) {
      return {};
    }
    const active = () => setIsActive(true);
    const inactive = () => setIsActive(false);
    return {
      onMouseEnter: e => {
        onMouseEnterProp?.(e);
        active();
      },
      onMouseLeave: e => {
        onMouseLeaveProp?.(e);
        inactive();
      },
      onDragLeave: e => {
        onDragLeaveProp?.(e);
        inactive();
      },
      onDragOver: e => {
        onDragOverProp?.(e);
        active();
      },
    };
  }, [disabled, onDragLeaveProp, onDragOverProp, onMouseEnterProp, onMouseLeaveProp]);
  return (
    <div className={classNames} ref={ref} {...resetProps} {...events}>
      <div className={stylex(sxDragAreaIcon)}>{icon}</div>
      <div className={stylex(sxDragAreaTitle, disabled && sxDisabledText)}>{dragTitle}</div>
      <p className={clsx(stylex(sxDragAreaDescription, disabled && sxDisabledText))}>{dragDescription}</p>
    </div>
  );
});

const InternalDragger: React.ForwardRefRenderFunction<UploadRef, DraggerUploadProps> = (
  { dragAreaConfig, ...restProps },
  ref,
) => (
  <Upload ref={ref} {...restProps} type="drag">
    {dragAreaConfig ? <UploadDragArea {...dragAreaConfig} /> : restProps.children}
  </Upload>
);

const Dragger = React.forwardRef(InternalDragger) as React.FC<DraggerUploadProps>;
if (__DEV__) {
  Dragger.displayName = 'Dragger';
}

export default Dragger;
