// @universe-design/react/src/upload/interface.tsx
import type React from 'react';
import type { StyleProps } from '../util/types';
import type { FileProps, PictureFileProps, UploadFile } from './uploadFile/interface';
import type {
  UploadItemRender,
  ShowUploadListInterface,
  UploadListType,
  UploadPictureFileListProps,
} from './uploadList/interface';
import type { UploadTriggerProps } from './UploadTrigger';

export type BeforeUploadValueType = void | boolean | string | Blob | File;

/**
 * @public
 */
export interface HttpRequestHeader {
  [key: string]: string;
}

/**
 * 上传类型：使用拖拽上传还是手动选择文件上传。
 * @public
 */
export type UploadType = 'drag' | 'select';

/**
 * 内部使用的文件类型。继承自 File, 拥有[File](https://developer.mozilla.org/en-US/docs/Web/API/File)的所有属性。
 * @public
 * @docgen
 */
export interface RcFile extends File {
  /**
   * 上次修改时间
   */
  readonly lastModifiedDate?: Date;
  /**
   * 唯一标识
   */
  uid: string;
}

/**
 * onChange 的参数
 * @public
 */
export interface UploadChangeParam<T extends object = UploadFile> {
  /**
   * 触发 change 的文件
   */
  file: T;
  /**
   * 最新文件列表
   */
  fileList: T[];
  /**
   * 上传进度事件
   */
  event?: { percent: number };
}

export interface UploadLocale {
  unknownSize: string;
  waiting: string;
  download: string;
  delete: string;
  preview: string;
  uploadAgain: string;
  failed: string;
  downloading: string;
  downloadFailed: string;
}

/**
 * @public
 * @docgen
 * @typeParam Response - Response
 */
export type ComposableUploadProps<Response = unknown> = {
  /**
   * 限制上传数量。当为 1 时，始终用最新上传的文件代替当前文件
   */
  maxCount?: number;
  /**
   * 当上传文件个数超过最大值时的回调
   */
  onExceed?: () => void;
  /**
   * 默认已经上传的文件列表
   */
  defaultFileList?: Array<UploadFile<Response>>;
  /**
   * 已经上传的文件列表（受控)
   */
  fileList?: Array<UploadFile<Response>>;
  /**
   * 文件列表发生变化
   */
  onChange?: (info: UploadChangeParam<UploadFile<Response>>) => void;
  /**
   * 自定义独一无二的 id
   */
  id?: string;
  /**
   * 是否禁用态
   * @defaultValue false
   */
  disabled?: boolean;
  /**
   * 自定义类名前缀
   */
  prefixCls?: string;
  /**
   * 上传调用的后端接口
   */
  action?: string | ((file: RcFile) => string) | ((file: RcFile) => PromiseLike<string>);

  /**
   * 上传所需额外参数,或返回上传额外参数的方法
   */
  // eslint-disable-next-line @typescript-eslint/ban-types
  data?: object | ((file: UploadFile<Response>) => object);
  /**
   * 上传接口使用的 HTTP Method
   */
  method?: 'POST' | 'PUT' | 'PATCH' | 'post' | 'put' | 'patch';
  /**
   * 提交数据时，自定义请求头部字段,IE10 以上有效
   */
  headers?: HttpRequestHeader;

  /**
   * 自己实现的request
   **/
  customRequest?: (options: UploadRequestOption) => { abort: () => void } | void;
  /**
   * 上传文件之前的钩子，参数为上传的文件，若返回 false 则停止上传。
   * 支持返回一个 Promise 对象，Promise 对象 reject 时则停止上传，
   * resolve 时开始上传（ resolve 传入 File 或 Blob 对象则上传 resolve 传入对象）；
   * 也可以返回 Upload.LIST_IGNORE，此时列表中将不展示此文件。 注意：IE9 不支持该方法
   * @param file
   * @param FileList
   */
  beforeUpload?: (file: RcFile, FileList: RcFile[]) => BeforeUploadValueType | Promise<BeforeUploadValueType>;
  /**
   * 提交数据时，文件所用字段名
   */
  name?: string;
  /**
   * 上传请求时是否携带 cookie
   */
  withCredentials?: boolean;
} & StyleProps;

/**
 * @public
 * @docgen
 * @typeParam Response - Response
 */
export type UploadProps<Response = unknown> = ComposableUploadProps<Response> &
  UploadTriggerProps &
  Pick<UploadPictureFileListProps, 'maxDisplayedPictureCount'> &
  Pick<PictureFileProps, 'previewFile'> &
  Pick<FileProps, 'iconRender' | 'progress'> & {
    /**
     * 自定义 preview 逻辑
     */
    onPreview?: (file: UploadFile) => void;
    /**
     * 点击下载文件时的回调，如果没有指定，则默认跳转到文件 url 对应的标签页
     */
    onDownload?: (file: UploadFile) => void;
    /**
     * 点击重试按钮的回调
     */
    onRetry?: (file: UploadFile) => void | boolean;
    /**
     * 点击移除文件时的回调，返回值为 false 时不移除。支持返回一个 Promise 对象，Promise 对象 resolve(false) 或 reject 时不移除
     */
    onRemove?: (file: UploadFile) => void | boolean | Promise<void | boolean>;
    /**
     * 上传说明区域, 请使用组合式用发来添加自定义区域
     */
    tips?: React.ReactNode;
    /**
     * 是否展示图片墙上传按钮（listType为picture才生效）
     * @defaultValue true
     */
    showPictureUploadButton?: boolean;
    /**
     * 自定义上传列表项
     */
    itemRender?: UploadItemRender;
    /**
     * 是否展示文件列表，及文件列表如何展示
     */
    showUploadList?: boolean | ShowUploadListInterface;
    /**
     * 文件列表形态
     */
    listType?: UploadListType;
  };

export type UploadRef = {
  upload: Uploader | null;
  fileList: UploadFile[];
  onProgress: (
    e: {
      percent: number;
      loaded: number;
    },
    file: RcFile,
  ) => void;
  onError: (error: Error, response: any, file: RcFile) => void;
  onSuccess: (response: any, file: RcFile, xhr: any) => void;
  onBatchStart: (
    batchFileInfoList: {
      file: RcFile;
      parsedFile: Exclude<BeforeUploadFileType, boolean>;
    }[],
  ) => void;
};

/**
 * 拖拽区域配置。
 * @public
 */
export interface DragAreaConfig {
  /**
   * 拖拽区域标题
   */
  dragTitle: React.ReactNode;
  /**
   * 拖拽区域 Icon
   */
  dragIcon?: React.ReactNode | ((isDisabled: boolean) => React.ReactNode);
  /**
   * 拖拽区域描述
   */
  dragDescription?: React.ReactNode;
}
/**
 * @public
 * @docgen
 */
export type DraggerUploadProps = UploadProps & {
  /**
   * 拖拽区域配置
   */
  dragAreaConfig?: DragAreaConfig;
};

export type BeforeUploadFileType = File | Blob | boolean | string;

export type Action = string | ((file: RcFile) => string | PromiseLike<string>);

/**
 * @public
 * @docgen
 */
export interface UploadProgressEvent extends ProgressEvent {
  /**
   * 进度百分百
   */
  percent: number;
}

export type UploadRequestMethod = 'POST' | 'PUT' | 'PATCH' | 'post' | 'put' | 'patch';

export type UploadRequestHeader = Record<string, string>;

export interface UploadRequestError extends Error {
  status?: number;
  method?: UploadRequestMethod;
  url?: string;
}

/**
 * @public
 */
export interface UploadRequestOption<T = any> {
  /**
   * 上传进度发生变化时，调用此方法更新进度
   */
  onProgress: (event: UploadProgressEvent) => void;
  /**
   * 上传失败时调用此方法更新文件状态
   */
  onError: (event: UploadRequestError | ProgressEvent, body?: T) => void;
  /**
   * 上传成功时调用此方法更新文件状态
   */
  onSuccess: (body?: T, xhr?: XMLHttpRequest) => void;
  /**
   * 同 uploadProps[data]
   */
  data?: Record<string, string | Blob>;
  /**
   * 调用上传接口时，文件字段的名称
   */
  filename?: string;
  /**
   * 要上传的文件
   */
  file: Exclude<BeforeUploadFileType, File | boolean> | RcFile;
  /**
   * 同[XMLHttpRequest.withCredentials](https://developer.mozilla.org/zh-CN/docs/Web/API/XMLHttpRequest/withCredentials)
   */
  withCredentials?: boolean;
  /**
   * 上传服务端地址
   */
  action: string;
  /**
   * 额外的请求头
   */
  headers?: UploadRequestHeader;
  /**
   * http method
   */
  method: UploadRequestMethod;
}

export type Uploader = {
  /**
   * 重试上传
   */
  retry: (file: RcFile) => void;
  /**
   * 中断上传
   */
  abort: (file: RcFile) => void;
  /**
   * 手动添加文件上传
   */
  uploadFiles: (files: File[]) => void;
};
