// @universe-design/react/src/upload/sx.ts
import stylex from '@universe-design/stylex';
// 拖拽区域
export const useDragAreaSx = (prefix: string) => {
  const sxDragArea = stylex.create(
    {
      cursor: 'pointer',
      border: `1px dashed ${$token['line-border-component']}`,
      paddingTop: 32,
      paddingBottom: 28,
      paddingLeft: 12,
      paddingRight: 12,
      textAlign: 'center',
      backgroundColor: $token['udtoken-component-outlined-bg'],
      borderRadius: $token['border-radius-md'],
    },
    `${prefix}-drag-area`,
  );

  const sxDragAreaDisabled = stylex.create(
    {
      cursor: 'not-allowed',
    },
    `${prefix}-drag-area--disabled`,
  );

  const sxDragAreaActive = stylex.create(
    {
      borderColor: $token['primary-content-default'],
    },
    `${prefix}-drag-area--active`,
  );

  const sxDragAreaIcon = stylex.create(
    {
      fontSize: 40,
      paddingBottom: 8,
      lineHeight: '1',
    },
    `${prefix}-drag-area-icon`,
  );

  const sxDragAreaTitle = stylex.create(
    {
      color: $token['text-title'],
      margin: 0,
      lineHeight: '22px',
      fontSize: 14,
    },
    `${prefix}-drag-area-title`,
  );

  const sxDragAreaDescription = stylex.create(
    {
      color: $token['text-placeholder'],
      margin: 0,
      lineHeight: '22px',
      fontSize: 14,
    },
    `${prefix}-drag-area-description`,
  );

  const sxDisabledText = stylex.create({
    color: $token['text-disabled'],
  });

  return {
    sxDragArea,
    sxDisabledText,
    sxDragAreaDescription,
    sxDragAreaTitle,
    sxDragAreaIcon,
    sxDragAreaDisabled,
    sxDragAreaActive,
  };
};
