// @universe-design/react/src/upload/UploadTrigger.tsx
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx } from '@universe-design/stylex/react';
import * as React from 'react';
import clsx from 'clsx';
import { v4 as uuid } from 'uuid';
import { useUploadContext } from './hooks/UploadContext';
import type { RcFile, UploadType } from './interface';
import traverseFileTree from './utils/traverse-file-tree';
import { forwardHTMLRef } from '../util/forwardRef';
import attrAccept from './utils/attr-accept';
import noop from 'lodash/noop';
import stylex from '@universe-design/stylex';

/**
 * @public
 * @docgen
 */
export type UploadTriggerProps = {
  /**
   * 是否允许同时选择多个文件
   */
  multiple?: boolean;
  /**
   * 是否允许上传文件夹（浏览器兼容有限）
   */
  directory?: boolean;
  /**
   * 允许的文件类型，同 input[type=file] 的 accept 字段
   */
  accept?: string;
  /**
   * 触发元素
   */
  children?: React.ReactNode;
  /**
   * 上传触发方式，选择文件或拖拽文件上传
   */
  type?: UploadType;
  /**
   * 是否点击时打开文件对话框
   * @defaultValue true
   */
  openFileDialogOnClick?: boolean;

  style?: React.CSSProperties;
  className?: string;
};

/**
 * 触发文件选择
 * 发送请求都在这个组件完成
 * @public
 */
export const UploadTrigger = forwardHTMLRef<UploadTriggerProps, 'div'>((props, ref) => {
  const {
    children,
    className,
    onClick,
    onKeyDown,
    openFileDialogOnClick = true,
    type = 'click',
    accept = '',
    directory,
    multiple = false,
    ...htmlProps
  } = props;
  const { rootPrefixCls, disabled, uploadFiles } = useUploadContext();
  const inputRef = React.useRef<HTMLInputElement>(null);

  const [uid, setUid] = React.useState(uuid());

  const onChange = React.useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => {
      const { files } = e.target;

      const acceptedFiles = [...(files ?? [])].filter((file: File) => !directory || attrAccept(file, accept || ''));
      uploadFiles(acceptedFiles);
      // reset
      setUid(uuid());
    },
    [accept, directory, uploadFiles],
  );

  // because input don't have directory/webkitdirectory type declaration
  const dirProps = directory ? { directory: 'directory', webkitdirectory: 'webkitdirectory' } : {};

  const handleClick: React.MouseEventHandler<HTMLDivElement> = React.useCallback(
    e => {
      inputRef.current?.click();
      onClick?.(e);
    },
    [onClick],
  );

  const handleKeyDown = React.useCallback(
    (e: React.KeyboardEvent<HTMLDivElement>) => {
      if (e.key === 'Enter') {
        inputRef.current?.click();
      }
      onKeyDown?.(e);
    },
    [onKeyDown],
  );

  const handleDrop = React.useCallback(
    e => {
      e.preventDefault();

      if (e.type === 'dragover') {
        return;
      }

      if (directory) {
        traverseFileTree(Array.prototype.slice.call(e.dataTransfer.items), uploadFiles, (_file: RcFile) =>
          attrAccept(_file, accept),
        );
      } else {
        const files = [...e.dataTransfer.files].filter((file: File) => attrAccept(file, accept));

        if (!multiple) {
          files.splice(1);
        }

        uploadFiles(files);
      }
    },
    [directory, uploadFiles, accept, multiple],
  );

  const events: React.HTMLAttributes<HTMLDivElement> = React.useMemo(() => {
    if (disabled) {
      return {};
    }

    return {
      onClick: openFileDialogOnClick ? handleClick : noop,
      onKeyDown: openFileDialogOnClick ? handleKeyDown : noop,
      onDrop: type === 'drag' ? handleDrop : noop,
      onDragOver: type === 'drag' ? handleDrop : noop,
      tabIndex: 0,
    };
  }, [disabled, handleClick, handleDrop, handleKeyDown, openFileDialogOnClick, type]);
  return (
    <div
      className={clsx(`${rootPrefixCls}-trigger`, className)}
      ref={ref}
      {...events}
      {...htmlProps}
      sx={[
        stylex.create({
          outline: 'none',
        }),
        type === 'click' &&
          stylex.create({
            display: 'inline-block',
          }),
      ]}
    >
      <input
        type="file"
        ref={inputRef}
        onClick={e => e.stopPropagation()}
        key={uid}
        style={{ display: 'none' }}
        accept={accept}
        {...dirProps}
        multiple={multiple}
        onChange={onChange}
      />
      {children}
    </div>
  );
});
