// @universe-design/react/src/upload/Upload.tsx
import * as React from 'react';
import stylex from '@universe-design/stylex';
import { ComposableUpload } from './ComposableUpload';
import { UploadCardFile, UploadPictureFile, UploadSimpleTextFile, UploadTextFile } from './uploadFile';
import { UploadCardFileList, UploadPictureFileList, UploadSimpleTextFileList, UploadTextFileList } from './uploadList';
import { PictureUploadTrigger } from './uploadList/PictureUploadTrigger';
import type { UploadTriggerProps } from './UploadTrigger';
import { UploadTrigger } from './UploadTrigger';
import { LIST_IGNORE } from './utils';
import type { FileProps, UploadFile } from './uploadFile/interface';
import type { UploadProps, UploadRef } from './interface';
import { useUploadContextSelector } from './hooks/UploadContext';
import { useUploadFile } from './uploadFile/useUploadFile';
import type { UploadItemRender, UploadListType } from './uploadList/interface';

export interface CompoundedComponent
  extends React.ForwardRefExoticComponent<React.PropsWithChildren<UploadProps> & React.RefAttributes<UploadRef>> {
  LIST_IGNORE: string;
}

const FileItem = ({
  itemRender,
  listType,
  fileProps,
}: {
  itemRender?: UploadItemRender;
  listType: UploadListType;
  fileProps: FileProps;
}) => {
  const { file } = fileProps;
  const fileList = useUploadContextSelector(ctx => ctx.fileList);
  let OriginComp = UploadTextFile;
  switch (listType) {
    case 'card': {
      OriginComp = UploadCardFile;
      break;
    }

    case 'picture': {
      OriginComp = UploadPictureFile;
      break;
    }
    case 'simple-text': {
      OriginComp = UploadSimpleTextFile;
      break;
    }
    default: {
      OriginComp = UploadTextFile;
      break;
    }
  }
  const fileCtx = useUploadFile(fileProps);
  const actions = {
    download: fileCtx.onDownload,
    preview: fileCtx.onPreview,
    remove: fileCtx.onRemove,
    retry: fileCtx.onRetry,
  };

  const originNode = <OriginComp {...fileProps} />;
  if (itemRender) {
    return <>{itemRender(originNode, file, fileList, actions)}</>;
  }
  return originNode;
};

/**
 *  推荐使用 `ComposableUpload` 来实现更灵活的组合式用法
 */
export const Upload = React.forwardRef<UploadRef, UploadProps>((props, ref) => {
  const {
    defaultFileList,
    fileList,
    disabled,
    onChange,
    maxCount,
    onExceed,
    children,
    listType = 'text',
    tips,
    showUploadList = true,
    showPictureUploadButton = true,
    className,
    style,
    // triggerProps start
    method,
    multiple,
    accept = listType === 'picture' ? 'image/*' : '',
    action,
    beforeUpload,
    customRequest,
    data,
    directory,
    headers,
    name,
    openFileDialogOnClick,
    type,
    withCredentials,
    // uploadList
    itemRender: itemRenderProp,
    maxDisplayedPictureCount,
    onPreview,
    onRetry,
    onRemove,
    onDownload,
    // triggerProps end
    ...uploadListProps
  } = props;

  const composableUploadProps = {
    defaultFileList,
    fileList,
    disabled,
    onChange,
    maxCount,
    onExceed,
    className,
    style,
    method,
    action,
    beforeUpload,
    customRequest,
    data,
    headers,
    name,
    openFileDialogOnClick,
    type,
    withCredentials,
  };

  const isPictureUpload = listType === 'picture';
  const uploadListIcons = typeof showUploadList !== 'object' ? {} : showUploadList;
  const uploadTriggerProps: UploadTriggerProps = {
    openFileDialogOnClick,
    type,
    accept,
    multiple,
    directory,
  };

  const getFileProps = (file: UploadFile) => ({
    ...uploadListProps,
    ...uploadListIcons,
    file,
    ...(onPreview ? { onPreview: () => onPreview(file) } : {}),
    ...(onRetry ? { onRetry: () => onRetry(file) } : {}),
    ...(onRemove ? { onRemove: () => onRemove(file) } : {}),
    ...(onDownload ? { onDownload: () => onDownload(file) } : {}),
  });

  const renderUploadList = () => {
    const classNames = children ? stylex(stylex.create({ marginTop: 8 })) : undefined;

    const itemRenderer = (file: UploadFile) => (
      <FileItem listType={listType} fileProps={getFileProps(file)} itemRender={itemRenderProp} />
    );
    switch (listType) {
      case 'card':
        return <UploadCardFileList className={classNames} itemRenderer={itemRenderer} />;
      case 'picture':
        return (
          <UploadPictureFileList
            maxDisplayedPictureCount={maxDisplayedPictureCount}
            uploadTrigger={showPictureUploadButton && <PictureUploadTrigger {...uploadTriggerProps} />}
            itemRenderer={itemRenderer}
          />
        );
      case 'simple-text':
        return <UploadSimpleTextFileList className={classNames} itemRenderer={itemRenderer} />;

      default:
        return <UploadTextFileList className={classNames} itemRenderer={itemRenderer} />;
    }
  };
  return (
    <ComposableUpload {...composableUploadProps} ref={ref}>
      {!isPictureUpload && <UploadTrigger {...uploadTriggerProps}>{children}</UploadTrigger>}
      {!isPictureUpload && tips}
      {showUploadList && renderUploadList()}
      {isPictureUpload && tips}
    </ComposableUpload>
  );
}) as CompoundedComponent;

Upload.LIST_IGNORE = LIST_IGNORE;

if (__DEV__) {
  Upload.displayName = 'Upload';
}
