// @universe-design/react/src/avatar/Avatar.tsx
import clsx from 'clsx';
import * as React from 'react';
import { useResizeObserver } from '@universe-design/react-utilities';
import { useSize } from '../hooks';
import { COMPONENT_VERSION_ID } from '../config-provider';
import type { PresetColorType, DefaultExtraSizeType, DefaultSizeType } from '../util';
import { isFunction, RegexChinese, RegexJapanese } from '../util';
import { useAvatarGroupContext } from './AvatarGroupContext';
import { useAvatarCssScope, UD_AVATAR_ID } from './common';
import type { TooltipProps } from '../tooltip';
import { Tooltip } from '../tooltip';
import composeRefs from '../util/composeRefs';
import { SquircleAvatar } from './squircle';

import './style';
import { getTextAvatarFontSizeBySize } from './utils';

/**
 * @privateMark
 * Avatar
 * 圆形+空心（群组头像，内外 border 都行，现在其实都是做成图了）（加载态没有 border）
 * 圆形+实心（个人头像，无 border，icon + 图片）（加载态没有 border）
 * 方形+实心（应用头像，内 border，icon + 图片 + 字）（加载态没有 border）
 *
 * Avatar.Mask
 * 头像蒙层，出现在文字下方，图片上方
 *
 * Avatar.Group
 * 重叠头像，需要 mask clip 镂空，中间 2px 的间隙
 * 并排头像，直接并排就行
 *
 */

/**
 * @public
 */
export type AvatarPropColor = PresetColorType;

const defaultSizeMap = {
  ['extra-small']: 20,
  ['small']: 24,
  ['medium']: 32,
  ['large']: 40,
  ['extra-large']: 48,
  ['default']: 32,
};

export function sizeNormalize(sizeProp: AvatarProps['size']) {
  let size = 0;
  if (typeof sizeProp === 'number') {
    size = sizeProp;
  } else if (!sizeProp || !defaultSizeMap[sizeProp]) {
    size = defaultSizeMap.default;
  } else {
    size = defaultSizeMap[sizeProp];
  }
  return size;
}

/**
 * @public
 */
export interface AvatarProps extends Omit<React.HTMLAttributes<HTMLSpanElement>, 'onLoad'> {
  key?: React.Key;
  /**
   * 头像形状
   * @defaultValue `circle`
   */
  shape?: 'square' | 'squircle' | 'circle';
  /**
   * 头像大小
   * @defaultValue `medium`
   */
  size?: DefaultSizeType | DefaultExtraSizeType | number;
  /**
   * 头字符类型距离左右两侧边界单位像素
   * @defaultValue 0
   */
  gap?: number;
  /**
   * 图片头像的资源地址或者图片元素
   */
  src?: React.ReactNode;
  /**
   * `img`标签`srcSet`属性
   */
  srcSet?: React.ImgHTMLAttributes<HTMLImageElement>['srcSet'];
  /**
   * `img`标签`draggable`属性
   */
  draggable?: React.ImgHTMLAttributes<HTMLImageElement>['draggable'];
  /**
   * `img`标签`alt`属性
   */
  alt?: React.ImgHTMLAttributes<HTMLImageElement>['alt'];
  /**
   * `img` 标签其他属性，可以用于添加 `loading` 属性。
   */
  imgProps?: Omit<React.ComponentPropsWithoutRef<'img'>, 'src' | 'srcSet' | 'draggable' | 'alt' | 'onLoad' | 'onError'>;
  /**
   * 设置头像的自定义图标
   */
  icon?: React.ReactNode;
  /**
   * `icon`的`renderProps`方法。`iconRenderer`第一个参数给用户传入一个建议的图标尺寸
   */
  iconRenderer?: (size: { width: number; height: number }) => React.ReactNode;
  /**
   * `style`属性透传
   */
  style?: React.CSSProperties;
  /**
   * `className`属性透传
   */
  className?: string;
  /**
   * 图片加载失败的事件，返回`false`会关闭组件默认的`fallback`行为
   */
  onError?: (event: React.SyntheticEvent<HTMLImageElement>) => boolean;
  /**
   * 图片加载成功的事件回调
   */
  onLoad?: React.ReactEventHandler<HTMLImageElement>;
  /**
   * @deprecated
   * 头像是否为有边框。由于 UD 的设计中 shape 和是否有描边是一一对应的，所以为了避免业务使用不当进行，我们将该逻辑内置，bordered 属性将会被移除
   * @defaultValue true
   */
  bordered?: boolean;
  /**
   * 头像颜色。我们只对 12 种预设颜色做了不同`colorStyle`的适配，如果用户需要自定义则可以覆盖`class`和`style`属性
   * @defaultValue `neutral`
   */
  color?: AvatarPropColor;
  /**
   * 头像的颜色样式，只对个人圆形头像生效
   * 1.`filled`填充，`color`属性为背景，`@static-white`为字色
   * 2.`outlined`轮廓，`color`属性为字色和边框色，背景为透明
   */
  colorStyle?: 'filled' | 'outlined';
  /**
   * 头像的显示文字。当提供`children`时，头像为字符类型。对于方块字（中、日）一行最多 2 个字符，对于拉丁系一行最多 4 个字符，最多显示两行
   */
  children?: React.ReactNode;
  /**
   * 头像气泡文字
   */
  tooltip?: TooltipProps;
  /**
   * @internal
   * 自动根据容器和内容缩放文字区域
   * @defaultValue true
   */
  autoScaleContent?: boolean;
}

export const Avatar = React.forwardRef<HTMLSpanElement, AvatarProps>(
  // eslint-disable-next-line max-lines-per-function
  (props, ref) => {
    const { size: ctxSize, shape: ctxShape } = useAvatarGroupContext();
    const {
      src,
      children,
      color = 'neutral',
      colorStyle = 'filled',
      gap = 0,
      className,
      style,
      icon,
      iconRenderer: iconRendererProps,
      srcSet,
      draggable,
      onError,
      onLoad,
      alt,
      tooltip,
      autoScaleContent = true,
      imgProps,
      shape = ctxShape,
      size: sizeProp = ctxSize,
      ...restProps
    } = props;

    const size = sizeNormalize(sizeProp);

    const [scale, setScale] = React.useState(1);
    const cssScope = useAvatarCssScope();
    const slicedText = [...(typeof children === 'string' ? children : '')].reduce(
      (pre, cur) => {
        const charStep = RegexChinese.test(cur) || RegexJapanese.test(cur) ? 2 : 1;
        if (pre.length + charStep <= 8) {
          if (pre.dist2lastLn + charStep > 4) {
            pre.result += `\n${cur}`;
            pre.dist2lastLn = charStep;
          } else {
            pre.result += cur;
            pre.dist2lastLn += charStep;
          }
          pre.length += charStep;
        }
        return pre;
      },
      {
        result: '',
        dist2lastLn: 0,
        length: 0,
      },
    ).result;
    const [isImgExist, setIsImgExist] = React.useState(true);
    React.useEffect(() => {
      setIsImgExist(true);
    }, [src]);
    const containerRef = React.useRef<HTMLSpanElement>(null);
    const [containerSizeRef, containerSize] = useSize<HTMLSpanElement>();
    React.useImperativeHandle<HTMLSpanElement | null, HTMLSpanElement | null>(ref, () => containerRef.current);
    const textRef = useResizeObserver<HTMLSpanElement>(
      entry => {
        const {
          contentRect: { width: textWidth, height: textHeight },
        } = entry;
        /**
         * 原型的相切矩形最大宽度
         * CW^2 = TW_max^2 + TH_max^2 = TW_max^2 * (1 + (TH / TW)^2)
         * TW_max = CW / SQRT(1 + (TH / TW)^2)
         */
        const containerWidth = containerRef.current?.clientHeight ?? 0;
        const textWidthMax = (containerWidth - gap * 2) / Math.sqrt(1 + (textHeight / textWidth) ** 2);
        if (textWidth > textWidthMax) {
          setScale(textWidthMax / textWidth);
        }
      },
      { disabled: !autoScaleContent },
    );

    let content: React.ReactNode;
    let isImageAvatar = false;
    let isTextAvatar = false;
    if (typeof src === 'string' && isImgExist) {
      isImageAvatar = true;
      const { className: imgClassNameProp, ...restImgProps } = imgProps ?? {};
      content = (
        <img
          className={clsx(cssScope('__image'), imgClassNameProp)}
          src={src}
          draggable={draggable}
          srcSet={srcSet}
          alt={alt}
          onLoad={onLoad}
          onError={ev => {
            const errorFlag = isFunction(onError) ? onError(ev) : undefined;
            if (errorFlag !== false) {
              setIsImgExist(false);
            }
          }}
          {...restImgProps}
        />
      );
    } else if (src && isImgExist) {
      isImageAvatar = true;
      content = src;
    } else if (icon || iconRendererProps) {
      // icon 不为空的话是 icon avatar
      const iconRenderer = iconRendererProps || (() => icon);
      content = (
        <span className={cssScope('__icon')}>
          {iconRenderer({
            width: containerSize.width / 2,
            height: containerSize.height / 2,
          })}
        </span>
      );
    } else {
      isTextAvatar = true;
      // slicedText 不为空的话是 text avatar
      content = (
        <span
          style={{
            // eslint-disable-next-line @typescript-eslint/no-magic-numbers
            fontSize: autoScaleContent ? 18 : getTextAvatarFontSizeBySize(size),
            transform: `translate(-50%, -50%) scale(${scale})`,
          }}
          className={cssScope('__text')}
          ref={textRef}
        >
          {typeof children === 'string' ? slicedText : children}
        </span>
      );
    }

    const mergedRef = React.useMemo(
      () => (isTextAvatar ? composeRefs(containerRef, containerSizeRef) : containerRef),
      [containerSizeRef, isTextAvatar],
    );

    let wrapperStyle: React.CSSProperties = {
      width: size,
      height: size,
    };
    const mergedClassName = clsx(
      cssScope(),
      {
        [cssScope(`-${color}`)]: !isImageAvatar && color,
        [cssScope(`-${colorStyle}`)]: shape === 'circle' ? colorStyle : colorStyle === 'filled',
        [cssScope(`-${shape}`)]: shape,
        [cssScope(`-${shape}-${size}`)]: shape === 'squircle',
        // [cssScope('-bordered')]: bordered,
        [cssScope('-image')]: isImageAvatar,
      },
      className,
    );
    if (shape === 'square') {
      wrapperStyle = {
        ...wrapperStyle,
      };
    }
    const avatar = (
      <span
        ref={mergedRef}
        style={{
          ...wrapperStyle,
          ...style,
        }}
        className={mergedClassName}
        {...restProps}
      >
        {content}
        {shape === 'squircle' && <SquircleAvatar bordered={isImageAvatar || color !== 'neutral'} size={size} />}
      </span>
    );

    if (tooltip) {
      return <Tooltip {...tooltip}>{avatar}</Tooltip>;
    }

    return avatar;
  },
);
if (__DEV__) {
  Avatar.displayName = 'Avatar';
}
Object.assign(Avatar, {
  [UD_AVATAR_ID]: true,
  [COMPONENT_VERSION_ID]: process.env.__VERSION__,
});
