// @universe-design/react/src/avatar/AvatarGroup.tsx
import clsx from 'clsx';
import * as React from 'react';
import stylex from '@universe-design/stylex';
import { COMPONENT_VERSION_ID } from '../config-provider';
import { isFunction, toFlatArray } from '../util';
import type { AvatarProps } from './Avatar';
import { Avatar, sizeNormalize } from './Avatar';
import { AvatarGroupContextProvider } from './AvatarGroupContext';
import { useAvatarCssScope, useAvatarGroupCssScope } from './common';
import { getAvatarGroupMoreContent } from './utils';
import { useGroupSvgMaskCls } from './AvatarGroupMask';
import { cssFontMedium } from '../stylex';

import './style';

/**
 * Avatar.Group 的属性
 * @public
 */
export interface AvatarGroupProps extends Pick<AvatarProps, 'size'> {
  /**
   * `className`属性透传。
   */
  className?: string;
  /**
   * 头像列表子节点。
   */
  children?: React.ReactNode;
  /**
   * `style`属性透传。
   */
  style?: React.CSSProperties;
  /**
   * 显示的最大头像个数
   * @defaultValue 5
   */
  maxCount?: number;
  /**
   * 自定义多余头像属性
   */
  maxProps?: AvatarProps;
  /**
   * 需要并列的头像列表。为了良好的`typescript`定义，以及避免用户不可预知的使用情况，建议选择直接传入`avatars`并让我们帮助用户渲染。
   */
  avatars?: AvatarProps[];
  /**
   * 头像之间的水平边距，默认为尺寸的 -0.25 倍。
   */
  spacing?: number;
  /**
   * 自定义多余头像渲染函数
   */
  maxRenderer?: (props: {
    maxAvatar: React.ReactNode;
    omittedAvatars: { node: React.ReactNode; index: number }[];
  }) => React.ReactNode;
  /**
   * 头像形状
   * @defaultValue `circle`
   */
  shape?: 'circle';
}

export const AvatarGroup: React.FC<AvatarGroupProps> = function AvatarGroup({
  maxCount = 5,
  children,
  avatars,
  shape = 'circle',
  size: sizeProp,
  spacing: spacingProp,
  maxProps = {},
  className,
  style,
  maxRenderer,
}) {
  const avatarCssScope = useAvatarCssScope();
  const size = sizeNormalize(sizeProp);
  const spacing = spacingProp ?? -size / 4;
  const cssScope = useAvatarGroupCssScope();
  const cls = clsx(cssScope(), className, spacing < 0 && cssScope('-stacked'));
  const hollowOutCls = useGroupSvgMaskCls(size, spacing);

  let childNodes: React.ReactNode[] = [];
  if (children) {
    childNodes = toFlatArray(children);
  } else if (avatars) {
    childNodes = avatars.map(avatar => (
      // eslint-disable-next-line react/jsx-key
      <Avatar {...avatar} />
    ));
  }

  function getAvatarsWithMargin(nodes: React.ReactNode[]): React.ReactNode[] {
    return nodes.map((child, index) => {
      if (React.isValidElement<{ style?: React.CSSProperties }>(child)) {
        return React.cloneElement(child, {
          key: child.key ?? `avatar-key-${index}`,
          style: {
            ...child.props.style,
            marginLeft: index === 0 ? undefined : spacing,
          },
        });
      }
      return child;
    });
  }

  if (childNodes.length > maxCount) {
    const children2Show = childNodes.slice(0, maxCount - 1);
    const children2Hide = childNodes.slice(maxCount - 1, childNodes.length);
    const { style: maxStyle, className: maxClassName, ...restMaxProps } = maxProps;
    const maxAvatarNode = (
      <Avatar
        key="avatar-key-max"
        style={{
          marginLeft: spacing,
          ...maxStyle,
        }}
        shape={shape}
        size={size}
        className={clsx(
          avatarCssScope('-max'),
          maxClassName,
          stylex(
            stylex.create({
              cursor: 'pointer',
              userSelect: 'none',
            }),
            cssFontMedium,
          ),
        )}
        autoScaleContent={false}
        {...restMaxProps}
      >
        {getAvatarGroupMoreContent({
          size,
          hideCount: children2Hide.length,
        })}
      </Avatar>
    );

    const maxNode = isFunction(maxRenderer)
      ? maxRenderer({
          maxAvatar: maxAvatarNode,
          omittedAvatars: children2Hide.map((node, index) => ({
            node,
            index: index + children2Show.length,
          })),
        })
      : maxAvatarNode;

    childNodes = [...getAvatarsWithMargin(children2Show), maxNode];
  } else {
    childNodes = getAvatarsWithMargin(childNodes);
  }

  childNodes.forEach((child, index) => {
    if (React.isValidElement<{ className?: string }>(child) && index < childNodes.length - 1) {
      childNodes[index] = React.cloneElement(child, {
        key: child.key,
        className: clsx(child.props.className, hollowOutCls),
      });
    }
  });

  return (
    <AvatarGroupContextProvider value={{ shape, size }}>
      <div className={cls} style={style}>
        {childNodes}
      </div>
    </AvatarGroupContextProvider>
  );
};

Object.assign(AvatarGroup, {
  [COMPONENT_VERSION_ID]: process.env.__VERSION__,
});
