// @universe-design/react/es/menu/internal-menu/SubMenu/InlineSubMenuList.js
import _extends from "@babel/runtime/helpers/esm/extends";
import * as React from 'react';
import CSSMotion from '@universe-design/rc-motion';
import { getMotion } from '../utils/motionUtil';
import MenuContextProvider, { RcMenuContext } from '../context/MenuContext';
import RcSubMenuList from './SubMenuList';
export default function InlineSubMenuList(_ref) {
  let {
    id,
    open,
    keyPath,
    children
  } = _ref;
  const fixedMode = 'inline';
  const {
    prefixCls,
    forceSubMenuRender,
    motion,
    defaultMotions,
    mode
  } = React.useContext(RcMenuContext);

  // Always use latest mode check
  const sameModeRef = React.useRef(false);
  sameModeRef.current = mode === fixedMode;

  // We record `destroy` mark here since when mode change from `inline` to others.
  // The inline list should remove when motion end.
  const [destroy, setDestroy] = React.useState(!sameModeRef.current);
  const mergedOpen = sameModeRef.current ? open : false;

  // ================================= Effect =================================
  // Reset destroy state when mode change back
  React.useEffect(() => {
    if (sameModeRef.current) {
      setDestroy(false);
    }
  }, [mode]);

  // ================================= Render =================================
  const mergedMotion = {
    ...getMotion(fixedMode, motion, defaultMotions)
  };

  // No need appear since nest inlineCollapse changed
  if (keyPath.length > 1) {
    mergedMotion.motionAppear = false;
  }

  // Hide inline list when mode changed and motion end
  const originOnVisibleChanged = mergedMotion.onVisibleChanged;
  mergedMotion.onVisibleChanged = newVisible => {
    if (!sameModeRef.current && !newVisible) {
      setDestroy(true);
    }
    return originOnVisibleChanged === null || originOnVisibleChanged === void 0 ? void 0 : originOnVisibleChanged(newVisible);
  };
  if (destroy) {
    return null;
  }
  return /*#__PURE__*/React.createElement(MenuContextProvider, {
    mode: fixedMode,
    locked: !sameModeRef.current
  }, /*#__PURE__*/React.createElement(CSSMotion, _extends({
    visible: mergedOpen
  }, mergedMotion, {
    forceRender: forceSubMenuRender,
    removeOnLeave: false,
    leavedClassName: `${prefixCls}-hidden`
  }), _ref2 => {
    let {
      className: motionClassName,
      style: motionStyle
    } = _ref2;
    return /*#__PURE__*/React.createElement(RcSubMenuList, {
      id: id,
      className: motionClassName,
      style: motionStyle
    }, children);
  }));
}
