// @universe-design/react/es/menu/internal-menu/SubMenu/PopupTrigger.js
import * as React from 'react';
import Trigger from '@universe-design/rc-trigger';
import clsx from 'clsx';
import raf from '@universe-design/rc-util/lib/raf';
import { RcMenuContext } from '../context/MenuContext';
import { rcPlacements, rcPlacementsRtl } from '../placements';
import { getMotion } from '../utils/motionUtil';
const popupPlacementMap = {
  horizontal: 'bottomLeft',
  vertical: 'rightTop',
  'vertical-left': 'rightTop',
  'vertical-right': 'leftTop'
};
export default function PopupTrigger(_ref) {
  let {
    prefixCls,
    visible,
    children,
    popup,
    popupClassName,
    popupOffset,
    disabled,
    mode,
    onVisibleChange
  } = _ref;
  const {
    getPopupContainer,
    rtl,
    subMenuOpenDelay,
    subMenuCloseDelay,
    builtinPlacements,
    triggerSubMenuAction,
    forceSubMenuRender,
    // Motion
    motion,
    defaultMotions
  } = React.useContext(RcMenuContext);
  const [innerVisible, setInnerVisible] = React.useState(false);
  const placement = rtl ? {
    ...rcPlacementsRtl,
    ...builtinPlacements
  } : {
    ...rcPlacements,
    ...builtinPlacements
  };
  const popupPlacement = popupPlacementMap[mode];
  const targetMotion = getMotion(mode, motion, defaultMotions);
  const mergedMotion = {
    ...targetMotion,
    leavedClassName: `${prefixCls}-hidden`,
    removeOnLeave: false,
    motionAppear: true
  };

  // Delay to change visible
  const visibleRef = React.useRef();
  React.useEffect(() => {
    visibleRef.current = raf(() => {
      setInnerVisible(visible);
    });
    return () => {
      raf.cancel(visibleRef.current);
    };
  }, [visible]);
  return /*#__PURE__*/React.createElement(Trigger, {
    prefixCls: prefixCls,
    popupClassName: clsx(`${prefixCls}-popup`, {
      [`${prefixCls}-rtl`]: rtl
    }, popupClassName),
    stretch: mode === 'horizontal' ? 'minWidth' : undefined,
    getPopupContainer: getPopupContainer,
    builtinPlacements: placement,
    popupPlacement: popupPlacement,
    popupVisible: innerVisible,
    popup: popup,
    popupAlign: popupOffset && {
      offset: popupOffset
    },
    action: disabled ? [] : triggerSubMenuAction && [triggerSubMenuAction],
    mouseEnterDelay: subMenuOpenDelay,
    mouseLeaveDelay: subMenuCloseDelay,
    onPopupVisibleChange: onVisibleChange,
    forceRender: forceSubMenuRender,
    popupMotion: mergedMotion
  }, children);
}
