// @byted-service/bytedtrace/lib/tracer/tracerConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TracerConfig = exports.ControlForComponent = void 0;
const tslib_1 = require("tslib");
const configProvider = tslib_1.__importStar(require("../confProvider"));
const CUSTOM_METRICS_LEVEL_OPEN = 0;
const SPAN_METRICS_LEVEL_SIMPLE = 1;
const SPAN_METRICS_LEVEL_CLOSE = 2;
class ControlForComponent {
    constructor(base, global) {
        this.base = base;
        this.global = global;
    }
    getByKey(key) {
        if (this.base && typeof this.base[key] === 'boolean') {
            return this.base[key];
        }
        if (this.global && typeof this.global[key] === 'boolean') {
            return this.global[key];
        }
        /* istanbul ignore next */
        return false;
    }
}
exports.ControlForComponent = ControlForComponent;
class TracerConfig {
    constructor(configurator) {
        this.configurator = configurator;
        this.config = {};
    }
    /**
     * 先拉取 tcc 配置然后初始化
     */
    async syncConfig() {
        const cf = await this.configurator.pullGlobalTracerConfig(true);
        this.init(cf);
    }
    /**
     * 初始化 tracer 时，先试用默认配置，然后从 tcc 拉取配置
     */
    asyncConfig() {
        this.init(configProvider.DEFAULT_GLOBAL_TRACER_CONFIG);
        this.syncConfig();
    }
    init(cf) {
        this.config.enableTrace = cf.enableTrace;
        this.config.enablePostTrace = cf.enablePostTrace;
        this.config.enableDyedHeader = cf.enableDyedHeader;
        this.config.enableSampling = cf.enableSampling;
        this.config.enableReportTotalSpan = cf.enableReportTotalSpan;
        this.config.spanMetricsLevel = cf.spanMetricsLevel;
        this.config.customMetricsLevel = cf.customMetricsLevel;
        this.config.lightModeByComponent = new ControlForComponent(cf.lightModeByComponent, cf.rootLightModeByComponent);
        this.config.disableThriftStyleMetricsByComponent =
            new ControlForComponent(cf.disableThriftStyleMetricsByComponent, cf.rootDisableThriftStyleMetricsByComponent);
    }
    enablePostTrace() {
        return this.config.enablePostTrace;
    }
    enableDyedHeader() {
        return this.config.enableDyedHeader;
    }
    enableSampling() {
        return this.config.enableSampling;
    }
    enableReportTotalSpan() {
        return this.config.enableReportTotalSpan;
    }
    enableTrace() {
        return this.config.enableTrace;
    }
    getSpanMetricsLevel() {
        return this.config.spanMetricsLevel;
    }
    isLightModeEnable(component) {
        return this.config.lightModeByComponent.getByKey(component);
    }
    isThriftStyleMetricsDisable(component) {
        return this.config.disableThriftStyleMetricsByComponent.getByKey(component);
    }
    isSimplifySpanMetrics() {
        return this.config.spanMetricsLevel === SPAN_METRICS_LEVEL_SIMPLE;
    }
    isMetricsEnable(custom) {
        if (custom) {
            return this.config.customMetricsLevel === CUSTOM_METRICS_LEVEL_OPEN;
        }
        return this.config.spanMetricsLevel !== SPAN_METRICS_LEVEL_CLOSE;
    }
}
exports.TracerConfig = TracerConfig;
