// @universe-design/react/src/date-picker/panels/selector/MonthYearSelector.tsx
import React from 'react';
import clsx from 'clsx';
import stylex from '@universe-design/stylex';
import DownBoldOutlined from '@universe-design/icons-react/DownBoldOutlined';
import type moment from 'moment';
import type { MODE } from '../../interface';
import { useCssScope, useLocale } from '../../../config-provider/hooks';
import { cssTypographyHeadline } from '../../../stylex';
import { Button } from '../../../button';
import { getYearRange } from '../base/utils';

export interface MonthYearSelectorProps {
  pickerValue: moment.Moment;
  mode: MODE;
  className?: string;
  collapsed: boolean;
  /**
   * 是否隐藏 header 中年月选择下拉箭头
   * @defaultValue false
   */
  hiddenCollapse?: boolean;
  onCollapsed?: (collapsed: boolean) => void;
  /**
   * 点击 header 年份时的回调
   */
  onYearClick?: () => void;
  /**
   * 点击 header 月份时的回调
   */
  onMonthClick?: () => void;
}

export const MonthYearSelector = (props: MonthYearSelectorProps) => {
  const { className, mode, collapsed, pickerValue, hiddenCollapse, onYearClick, onMonthClick, onCollapsed } = props;

  const {
    locale: localeName,
    components: { datePicker: locale },
  } = useLocale();
  const cssScope = useCssScope('picker');
  const headerPrefixCls = cssScope('-panel-header');

  const wrapYear = (children: string) => (
    <span key="year" className={`${headerPrefixCls}-btn`} onClick={onYearClick}>
      {children}
    </span>
  );
  const timeNodes = locale.monthYearSelector.f({
    year: yearFormat => wrapYear(pickerValue.clone().locale(localeName).format(yearFormat.toString())),
    month: monthFormat => (
      <span key="month" onClick={onMonthClick} className={`${headerPrefixCls}-btn`}>
        {pickerValue.clone().locale(localeName).format(monthFormat.toString())}
      </span>
    ),
  });

  const handleCollapsed: React.MouseEventHandler = e => {
    e.stopPropagation();
    e.preventDefault();
    onCollapsed?.(!collapsed);
  };

  return (
    <div className={clsx(`${headerPrefixCls}-content`, className, stylex(cssTypographyHeadline))}>
      {mode === 'year' ? wrapYear(getYearRange(pickerValue)) : timeNodes}
      {!hiddenCollapse && (
        <Button
          type="icon"
          className={clsx(`${headerPrefixCls}-icon`, `${headerPrefixCls}-collapse`)}
          onClick={handleCollapsed}
          data-testid="ud__picker-collapse"
          size="extra-small"
          icon={
            <DownBoldOutlined
              className={clsx({
                [`${headerPrefixCls}-rotate`]: true,
                [`${headerPrefixCls}-rotate180`]: !collapsed,
              })}
            />
          }
        />
      )}
    </div>
  );
};
