// @byted-service/bytedtrace/lib/serializer/compressor.js
"use strict";
/* istanbul ignore file */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCompressors = exports.registerCompressor = exports.globalCompressor = exports.Compressors = exports.noneCompressor = exports.None = void 0;
const tslib_1 = require("tslib");
const consts = tslib_1.__importStar(require("./consts"));
/**
 * 目前 sdk 侧还没有使用压缩，可以先忽略
 * 暂时先 ignore
 */
exports.None = 0;
exports.noneCompressor = {
    compress(data) {
        return Buffer.from(data);
    },
    deCompress(data) {
        return data;
    },
    copy() {
        return Object.assign({}, this);
    },
};
class Compressors {
    constructor(compressors) {
        this.compressors = compressors || {};
    }
    registerCompressor(compressorType, compressor) {
        this.compressors[compressorType] = compressor;
    }
    compress(compressorType, data) {
        if (!this.compressors[compressorType]) {
            throw new Error(consts.CAN_NOT_FIND_COMPRESSOR_ERROR);
        }
        return this.compressors[compressorType].compress(data);
    }
    deCompress(compressorType, data) {
        if (!this.compressors[compressorType]) {
            throw new Error(consts.CAN_NOT_FIND_COMPRESSOR_ERROR);
        }
        return this.compressors[compressorType].deCompress(data);
    }
}
exports.Compressors = Compressors;
const cps = { 0: exports.noneCompressor };
exports.globalCompressor = {
    compressors: new Compressors(cps),
};
const registerCompressor = (compressorType, compressor) => {
    exports.globalCompressor.compressors.registerCompressor(compressorType, compressor);
};
exports.registerCompressor = registerCompressor;
const getCompressors = () => {
    const cps = {};
    for (const key in exports.globalCompressor.compressors.compressors) {
        if (exports.globalCompressor.compressors.compressors.hasOwnProperty(key)) {
            cps[key] = exports.globalCompressor.compressors.compressors[key].copy();
        }
    }
    return new Compressors(cps);
};
exports.getCompressors = getCompressors;
