// @byted-service/bytedtrace/lib/serializer/consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OVER_DICTIONARY_SIZE_ERROR = exports.FLATTEN_DICTIONARY_ERROR = exports.DICTIONARY_HAS_NOT_KEY_ERROR = exports.MAX_DICTIONARY_SIZE = exports.CAN_NOT_FIND_COMPRESSOR_ERROR = exports.DEFAULT_SERIALIZER_HEADERS = exports.DEFAULT_USE_DIC = exports.DEFAULT_BIG_ENDIAN = exports.INCOMPLETE_RESERVE_LOG_HEADER_ERROR = exports.EMPTY_RESERVE_LOG_HEADER_ERROR = exports.DEFAULT_CONTENT_COMPRESSION = exports.DEFAULT_USER_ID = exports.DEFAULT_RESERVED = exports.DEFAULT_COMPRESSION = exports.DEFAULT_PROTO_TYPE = exports.DEFAULT_VERSION = exports.EMPTY_STRING = exports.EMPTY_LENGTH_BYTES = exports.EMPTY_HEADER_ERROR = exports.FLOW_DE_SERIALIZER_TYPE = exports.FLOW_SERIALIZER_TYPE = exports.FALSE_BOOL_DATA = exports.TRUE_BOOL_DATA = exports.UN_RECOGNIZE_TYPE_ERROR = exports.INDEX_OUT_OF_BOUND_ERROR = exports.UUID_TYPE = exports.BYTES_TYPE = exports.TEXT_TYPE = exports.IPV6_TYPE = exports.IPV4_TYPE = exports.DOUBLE_TYPE = exports.UINT64_TYPE = exports.LONG_TYPE = exports.INT_TYPE = exports.BOOL_TYPE = exports.STRING_TYPE = exports.TRUE_BYTE = exports.FALSE_BYTE = exports.STRING_SPLIT_BYTE = exports.MAX_STRING_LENGTH = void 0;
exports.MAX_STRING_LENGTH = 255;
exports.STRING_SPLIT_BYTE = 0x00;
exports.FALSE_BYTE = 0x0;
exports.TRUE_BYTE = 0x1;
exports.STRING_TYPE = 0; // 长度小于255的string
exports.BOOL_TYPE = 1;
exports.INT_TYPE = 2; // 4字节, 8,16,32位有符号,byte,8,16位无符号
exports.LONG_TYPE = 3; // 8字节, 64位有符号,int
exports.UINT64_TYPE = 4; // 8字节, 32,64位无符号
exports.DOUBLE_TYPE = 5; // 8字节, float32,float64
exports.IPV4_TYPE = 21;
exports.IPV6_TYPE = 22;
exports.TEXT_TYPE = 23; // 长度大于255的string
exports.BYTES_TYPE = 24; // 二进制字符流
exports.UUID_TYPE = 25; // 128位
exports.INDEX_OUT_OF_BOUND_ERROR = 'Index out of bound!';
exports.UN_RECOGNIZE_TYPE_ERROR = 'type is unRecognize!';
exports.TRUE_BOOL_DATA = Buffer.from([exports.BOOL_TYPE, exports.TRUE_BYTE]);
exports.FALSE_BOOL_DATA = Buffer.from([exports.BOOL_TYPE, exports.FALSE_BYTE]);
// const notHeaderOptionFlowError = 'not headerOptionFlow';
exports.FLOW_SERIALIZER_TYPE = 'flow';
exports.FLOW_DE_SERIALIZER_TYPE = 'flow';
exports.EMPTY_HEADER_ERROR = 'log header is empty';
exports.EMPTY_LENGTH_BYTES = Buffer.alloc(4);
exports.EMPTY_STRING = '';
exports.DEFAULT_VERSION = 0;
exports.DEFAULT_PROTO_TYPE = 0;
exports.DEFAULT_COMPRESSION = 0;
exports.DEFAULT_RESERVED = 0;
exports.DEFAULT_USER_ID = '0';
exports.DEFAULT_CONTENT_COMPRESSION = 0;
exports.EMPTY_RESERVE_LOG_HEADER_ERROR = 'reserve header is empty';
exports.INCOMPLETE_RESERVE_LOG_HEADER_ERROR = 'timestamp is require for reserve header';
exports.DEFAULT_BIG_ENDIAN = false;
exports.DEFAULT_USE_DIC = true;
exports.DEFAULT_SERIALIZER_HEADERS = {
    version: exports.DEFAULT_VERSION,
    protoType: exports.DEFAULT_PROTO_TYPE,
    compression: exports.DEFAULT_COMPRESSION,
    reserved: exports.DEFAULT_RESERVED,
    contentCompression: exports.DEFAULT_CONTENT_COMPRESSION,
    userId: exports.DEFAULT_USER_ID
};
exports.CAN_NOT_FIND_COMPRESSOR_ERROR = '[bytedTrace] can not find such compressor';
exports.MAX_DICTIONARY_SIZE = 255;
exports.DICTIONARY_HAS_NOT_KEY_ERROR = "dictionary doesn't have such key";
exports.FLATTEN_DICTIONARY_ERROR = 'dictionary flatten error';
exports.OVER_DICTIONARY_SIZE_ERROR = 'key number is not allowed over than 256';
