// @byted-service/bytedtrace/lib/serializer/serializer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Serializer = void 0;
const tslib_1 = require("tslib");
/**
 * 序列化器，挂载一个 buffer 数据，写入数据暂存在这个 buffer
 * 调用 serialize 方法后会按协议要求序列化 buffer 中的数据并返回
 */
const utils = tslib_1.__importStar(require("./utils"));
const consts = tslib_1.__importStar(require("./consts"));
const byteWriteBuffer_1 = require("./byteWriteBuffer");
const compressor_1 = require("./compressor");
const logger_1 = require("../logger");
const dictionary_1 = require("./dictionary");
const types_1 = require("./types");
const utils_1 = require("../utils");
const consts_1 = require("../consts");
class Serializer extends types_1.ISerializer {
    constructor(serializerType = consts.FLOW_SERIALIZER_TYPE, opt = {}) {
        super();
        if (serializerType === consts.FLOW_SERIALIZER_TYPE) {
            const bigEndian = typeof opt.bigEndian === 'boolean'
                ? opt.bigEndian
                : consts.DEFAULT_BIG_ENDIAN;
            const useDic = typeof opt.useDic === 'boolean'
                ? opt.useDic
                : consts.DEFAULT_USE_DIC;
            this.bigEndian = bigEndian;
            this.useDic = useDic;
            this.buffer = new byteWriteBuffer_1.ByteWriteBuffer(bigEndian, opt.logger);
            this.tempBuffer = new byteWriteBuffer_1.ByteWriteBuffer(bigEndian, opt.logger);
            this.contentSize = 0;
            this.dictionary = new dictionary_1.Dictionary();
            this.dictionary.reset();
            this.compressors = (0, compressor_1.getCompressors)();
            this.totalLengthPos = 4;
            this.headerRegionCompressedLengthPos = 8;
            this.headerRegionOriginLengthPos = 12;
            this.dataPacks = [];
            this.contentHeader = Buffer.alloc(9);
        }
        this.logger = opt.logger;
        logger_1.debugLogger.debug('constructor this.bigEndian: ', this.bigEndian);
    }
    setHeader(opt) {
        opt = this.fixHeader(opt);
        this.contentHeader.fill(0);
        this.contentHeader[8] = opt.contentCompression;
        // client 端暂不使用压缩
        opt.contentCompression = 0;
        opt.compression = 0;
        this.headerOption = opt;
        this.genHeader(true);
    }
    fixHeader(opt) {
        return (0, utils_1.transPropertyToTarget)(opt, {
            version: consts.DEFAULT_SERIALIZER_HEADERS.version,
            protoType: consts.DEFAULT_SERIALIZER_HEADERS.protoType,
            compression: consts.DEFAULT_SERIALIZER_HEADERS.compression,
            reserved: consts.DEFAULT_SERIALIZER_HEADERS.reserved,
            contentCompression: consts.DEFAULT_SERIALIZER_HEADERS.contentCompression,
            userId: consts.DEFAULT_SERIALIZER_HEADERS.userId,
        });
    }
    setCommonHeaders(headers) {
        this.commonHeaders = headers;
        return this.genCommonHeader(true);
    }
    genCommonHeader(force) {
        if (this.commonHeaderByte === undefined || force) {
            if (this.commonHeaders === undefined) {
                this.commonHeaderByte = Buffer.alloc(4);
            }
            else {
                const bf = this.getTempBuffer();
                let size = 0;
                bf.write(consts.EMPTY_LENGTH_BYTES);
                if (this.useDic) {
                    this.commonHeaders.forEach((kv) => {
                        this.dictionary.addCommonKey(kv.key);
                        const id = this.dictionary.coding(kv.key);
                        const l = bf.writeDicKeyValue(id, kv.value, kv.type);
                        size += l;
                    });
                }
                else {
                    this.commonHeaders.forEach((kv) => {
                        const l = bf.writeKeyValue(kv.key, kv.value, kv.type);
                        size += l;
                    });
                }
                bf.setPosValue(0, size);
                this.commonHeaderByte = bf.bytesCopy();
                /* istanbul ignore if */
                if (consts_1.IS_DEBUG) {
                    logger_1.debugLogger.debug('genCommonHeader size: %d', size);
                    logger_1.debugLogger.debug('genCommonHeader this.commonHeaderByte: %s', utils.bufferToInt(this.commonHeaderByte));
                }
            }
        }
    }
    /**
     * 填充 header 共 24 Byte
     * @param force
     * @returns
     */
    genHeader(force) {
        if (!this.headerOption) {
            return;
        }
        if (force || !this.header) {
            // 清空数据
            this.header = this.header || Buffer.alloc(24);
            const { header, headerOption } = this;
            header.fill(0);
            header[0] = headerOption.version;
            header[1] = headerOption.protoType;
            header[2] = headerOption.compression;
            header[3] = headerOption.reserved;
            // 这里还不能确定中间 12 Byte 的具体填充值，分别为
            // 1、totalLength 4B
            // 2、compressedLength 4B
            // 3、originLength 4B
            const ids = utils.int64ToBytes(headerOption.userId, this.bigEndian);
            utils.appendToTarget(header, ids, 16);
            logger_1.debugLogger.debug('genHeader this.header: %o', header);
        }
    }
    feed(dataPack) {
        if (this.useDic) {
            if (dataPack.logHeader && dataPack.logHeader.customizeHeader) {
                dataPack.logHeader.customizeHeader.forEach((kv) => this.dictionary.addKey(kv.key));
            }
            if (dataPack.logContent && dataPack.logContent.keyValues) {
                dataPack.logContent.keyValues.forEach((kv) => this.dictionary.addKey(kv.key));
            }
        }
        this.dataPacks.push(dataPack);
    }
    serialize() {
        if (this.dataPacks === undefined || this.dataPacks.length <= 0) {
            return;
        }
        if (!this.headerOption) {
            throw new Error(consts.EMPTY_HEADER_ERROR);
        }
        this.buffer.reset();
        this.buffer.write(this.header);
        // 写入头部的位移
        const pos = this.buffer.len();
        let buffer;
        // client 暂不使用压缩
        /* istanbul ignore if */
        if (this.headerOption.compression !== compressor_1.None) {
            // 有压缩器则使用新的 buffer，写完后压缩，再写入 this.buffer，否则直接将数据写入 this.buffer
            buffer = this.getTempBuffer();
        }
        else {
            buffer = this.buffer;
        }
        this.patternLengthPos = buffer.len();
        this.buffer.write(consts.EMPTY_LENGTH_BYTES);
        if (this.useDic) {
            const dictionary = this.dictionary.flatten();
            let patternLength = 0;
            dictionary.forEach((str, _) => {
                patternLength += buffer.writeStringKey(str);
            });
            buffer.setPosValue(this.patternLengthPos, patternLength);
        }
        this.genCommonHeader(false);
        logger_1.debugLogger.debug('serialize this.commonHeaderByte: %o', this.commonHeaderByte);
        buffer.write(this.commonHeaderByte);
        this.logHeaderLengthPos = buffer.len();
        buffer.write(consts.EMPTY_LENGTH_BYTES);
        let totalHeaderLength = 0;
        /* istanbul ignore if */
        if (consts_1.IS_DEBUG) {
            logger_1.debugLogger.debug('serialize this.dataPacks: %s', JSON.stringify(this.dataPacks));
        }
        this.dataPacks.forEach((dp) => {
            const headers = dp.logHeader;
            if (headers) {
                const pos = buffer.len();
                let headerWriteLength = 0;
                const wl = buffer.write(consts.EMPTY_LENGTH_BYTES);
                const reserveHeader = headers.reserveHeader;
                /* istanbul ignore if */
                if (!reserveHeader) {
                    throw new Error(consts.EMPTY_RESERVE_LOG_HEADER_ERROR);
                }
                /* istanbul ignore if */
                if (typeof reserveHeader.timestamp !== 'number') {
                    throw new Error(consts.INCOMPLETE_RESERVE_LOG_HEADER_ERROR);
                }
                ['source', 'context', 'logId'].forEach((e) => (reserveHeader[e] =
                    reserveHeader[e] || consts.EMPTY_STRING));
                headerWriteLength +=
                    buffer.writeValue(reserveHeader.timestamp, 'uint64') +
                        buffer.writeValue(reserveHeader.source, 'string') +
                        buffer.writeValue(reserveHeader.context, 'string') +
                        buffer.writeValue(reserveHeader.logId, 'string');
                if (headers.customizeHeader) {
                    if (this.useDic) {
                        headers.customizeHeader.forEach((kv) => {
                            const id = this.dictionary.coding(kv.key);
                            headerWriteLength += buffer.writeDicKeyValue(id, kv.value, kv.type);
                        });
                    }
                    else {
                        headers.customizeHeader.forEach((kv) => {
                            headerWriteLength += buffer.writeKeyValue(kv.key, kv.value, kv.type);
                        });
                    }
                }
                buffer.setPosValue(pos, headerWriteLength);
                totalHeaderLength += wl + headerWriteLength;
            }
        });
        buffer.setPosValue(this.logHeaderLengthPos, totalHeaderLength);
        // client 端暂不使用压缩
        /* istanbul ignore if */
        if (this.headerOption.compression !== compressor_1.None) {
            const headerData = buffer.bytesCopy();
            const rs = this.compressors.compress(this.headerOption.compression, headerData);
            const wl = this.buffer.write(rs);
            this.buffer.setPosValue(this.headerRegionCompressedLengthPos, wl);
            this.buffer.setPosValue(this.headerRegionOriginLengthPos, headerData.length);
        }
        else {
            const headerLength = this.buffer.len() - pos;
            this.buffer.setPosValue(this.headerRegionCompressedLengthPos, headerLength);
            this.buffer.setPosValue(this.headerRegionOriginLengthPos, headerLength);
        }
        this.logContentCompressedLengthPos = this.buffer.len();
        this.logContentOriginLengthPos = this.logContentCompressedLengthPos + 4;
        this.buffer.write(this.contentHeader);
        const contentPos = this.buffer.len();
        // client 端暂不使用压缩
        /* istanbul ignore if */
        if (this.headerOption.contentCompression !== compressor_1.None) {
            buffer = this.getTempBuffer();
        }
        else {
            buffer = this.buffer;
        }
        this.dataPacks.forEach((dataPack) => {
            const content = dataPack.logContent;
            if (content) {
                let contentWriteLength = 0;
                const pos = buffer.len();
                buffer.write(consts.EMPTY_LENGTH_BYTES);
                if (this.useDic) {
                    content.keyValues.forEach((kv) => {
                        const id = this.dictionary.coding(kv.key);
                        contentWriteLength += buffer.writeDicKeyValue(id, kv.value, kv.type);
                    });
                }
                else {
                    content.keyValues.forEach((kv) => {
                        contentWriteLength += buffer.writeKeyValue(kv.key, kv.value, kv.type);
                    });
                }
                buffer.setPosValue(pos, contentWriteLength);
            }
        });
        // client 端暂不使用压缩
        /* istanbul ignore if */
        if (this.headerOption.contentCompression !== compressor_1.None) {
            const contentData = buffer.bytesCopy();
            const rs = this.compressors.compress(this.headerOption.contentCompression, contentData);
            const wl = this.buffer.write(rs);
            this.buffer.setPosValue(this.logContentCompressedLengthPos, wl);
            this.buffer.setPosValue(this.logContentOriginLengthPos, contentData.length);
        }
        else {
            const contentLength = this.buffer.len() - contentPos;
            this.buffer.setPosValue(this.logContentCompressedLengthPos, contentLength);
            this.buffer.setPosValue(this.logContentOriginLengthPos, contentLength);
        }
        const totalLength = this.buffer.len() - pos;
        this.buffer.setPosValue(this.totalLengthPos, totalLength);
        return this.buffer.bytesCopy();
    }
    clear() {
        this.dataPacks = [];
        this.dictionary.clear();
    }
    reset() {
        this.dataPacks = [];
        this.commonHeaderByte = undefined;
        this.commonHeaders = undefined;
        this.dictionary.reset();
    }
    getTempBuffer() {
        this.tempBuffer.reset();
        return this.tempBuffer;
    }
}
exports.Serializer = Serializer;
