// @universe-design/react/lib/metrics/MetricsHeader.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MetricsHeader = void 0;
var _u1rkqst = _interopRequireDefault(require("./../.stylex/u1rkqst1"));
var _ujhu89j = _interopRequireDefault(require("./../.stylex/ujhu89j"));
var _u1g5q35g = _interopRequireDefault(require("./../.stylex/u1g5q35g"));
var _u13d3rh = _interopRequireDefault(require("./../.stylex/u13d3rh8"));
var _ukchsq = _interopRequireDefault(require("./../.stylex/ukchsq7"));
var _u1gmpn6f = _interopRequireDefault(require("./../.stylex/u1gmpn6f"));
var _umue7z = _interopRequireDefault(require("./../.stylex/umue7z"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = require("@universe-design/stylex/react");
var _InfoOutlined = _interopRequireDefault(require("@universe-design/icons-react/InfoOutlined"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _tooltip = _interopRequireDefault(require("../tooltip"));
var _text = require("../text");
var _forwardRef = require("../util/forwardRef");
var _MetricsContext = require("./MetricsContext");
var _slot = require("../shared/slot");
var _MetricsGroup = require("./MetricsGroup");
var _useMetrics = require("./useMetrics");
var _reactUtilities = require("@universe-design/react-utilities");
var React = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
require("./style");
//.u1fssmi4{color:@icon-n2;}
//.u1nvqvck{font-size:calc(1em + 2px);}
//.uqkj31s{vertical-align:calc(0.36 * 1em - 0.5 * (1em + 2px));}
//.uly5x3f{width:100%;}
//.u22iagw{display:flex;}
//.u1guyajx{align-items:baseline;}
//.u4xv25i{gap:4px;}
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const useMetricsHeaderStylexStyles = (0, _react.makeStylexStyles)(_ref => {
  let {
    prefix
  } = _ref;
  return {
    header: {
      "props": {
        "a9b677": _u13d3rh.default,
        "mc9l5x": _ukchsq.default,
        "Bt984gj": _u1gmpn6f.default,
        "rmohyg": _umue7z.default
      },
      "alias": `${prefix}__header`
    },
    icon: {
      "props": {
        "sj55zd": _u1rkqst.default,
        "Be2twd7": _ujhu89j.default,
        "ha4doy": _u1g5q35g.default
      },
      "alias": `${prefix}__icon`
    }
  };
});
const MetricsHeader = exports.MetricsHeader = (0, _forwardRef.forwardHTMLRef)((props, ref) => {
  const {
    mainTitle,
    mainInfoIcon,
    mainHeaderContentAfter,
    className,
    ...restProps
  } = props;
  const {
    rootCssScope,
    size
  } = (0, _MetricsContext.useMetricsContext)();
  const {
    index,
    register
  } = (0, _useMetrics.useMetricsHeaderDescendant)();
  const recalculateHeaderHeights = (0, _MetricsGroup.useMetricsGroupContextSelector)(context => context === null || context === void 0 ? void 0 : context.recalculateHeaderHeights);
  const headerRowMaxHeight = (0, _MetricsGroup.useMetricsGroupContextSelector)(context => {
    var _context$headerRowInf, _context$headerRowInf2;
    return context === null || context === void 0 ? void 0 : (_context$headerRowInf = context.headerRowInfo) === null || _context$headerRowInf === void 0 ? void 0 : (_context$headerRowInf2 = _context$headerRowInf[index]) === null || _context$headerRowInf2 === void 0 ? void 0 : _context$headerRowInf2.maxHeight;
  });
  const titleLineClamp = (0, _MetricsGroup.useMetricsGroupContextSelector)(context => context === null || context === void 0 ? void 0 : context.titleLineClamp) || 2;
  const styles = useMetricsHeaderStylexStyles({
    prefix: rootCssScope()
  });
  const mainTitleProps = (0, _slot.resolveShorthandOld)(mainTitle);
  const mainInfoIconProps = (0, _slot.resolveShorthandOld)(mainInfoIcon);
  const mainHeaderContentAfterProps = (0, _slot.resolveShorthandOld)(mainHeaderContentAfter);
  const {
    tooltip
  } = mainInfoIconProps || {};
  return (0, _react.jsx)(React.Fragment, null, (0, _react.jsx)("div", (0, _extends2.default)({
    className: (0, _clsx.default)(`${rootCssScope()}__header__wrapper`, className),
    style: {
      height: headerRowMaxHeight
    }
  }, restProps), (0, _react.jsx)(_reactUtilities.ResizeObserver, {
    onResize: recalculateHeaderHeights
  }, (0, _react.jsx)(_text.Text, {
    as: "div",
    ref: (0, _reactUtilities.useMergedRefs)(ref, register),
    sx: [styles.header],
    variant: size === 'medium' ? 'body0' : 'title5'
  }, mainTitle !== null && (0, _react.jsx)(_text.Text, {
    lineClamp: titleLineClamp
  }, mainTitleProps === null || mainTitleProps === void 0 ? void 0 : mainTitleProps.children), mainInfoIcon !== null && (0, _react.jsx)(_tooltip.default, tooltip, (0, _react.jsx)("span", null, (0, _react.jsx)(_InfoOutlined.default, {
    className: (0, _stylex.default)(styles.icon),
    width: 1
  }))), mainHeaderContentAfter !== null && (0, _react.jsx)("div", null, mainHeaderContentAfterProps === null || mainHeaderContentAfterProps === void 0 ? void 0 : mainHeaderContentAfterProps.children)))));
});
