// @universe-design/react/lib/metrics/MetricsContext.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useMetricsContext = exports.MetricsProvider = void 0;
var _reactProvider = require("@universe-design/react-provider");
var _react = _interopRequireWildcard(require("react"));
var React = _react;
var _configProvider = require("../config-provider");
const {
  useContext: useMetricsContext,
  Provider: MetricsContextProvider
} = (0, _reactProvider.createContextHook)('MetricsContext', {
  strict: true,
  customizedContextMethods: {
    createContext: _react.createContext,
    useContext: _react.useContext
  }
});
exports.useMetricsContext = useMetricsContext;
const MetricsProvider = _ref => {
  let {
    size = 'medium',
    position = 'afterAll',
    children
  } = _ref;
  const rootCssScope = (0, _configProvider.useCssScope)('metrics');
  return /*#__PURE__*/React.createElement(MetricsContextProvider, {
    value: {
      rootCssScope,
      size,
      position
    }
  }, children);
};
exports.MetricsProvider = MetricsProvider;
