// @universe-design/react/lib/metrics/MetricsRoot.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MetricsRoot = void 0;
var _u40u48c = _interopRequireDefault(require("./../.stylex/u40u48c"));
var _umue7z = _interopRequireDefault(require("./../.stylex/umue7z"));
var _uohwdt = _interopRequireDefault(require("./../.stylex/uohwdt"));
var _u1d70x8k = _interopRequireDefault(require("./../.stylex/u1d70x8k"));
var _u1cpp6s = _interopRequireDefault(require("./../.stylex/u1cpp6s6"));
var _uyretdy = _interopRequireDefault(require("./../.stylex/uyretdy"));
var _unqnxb = _interopRequireDefault(require("./../.stylex/unqnxb"));
var _ukchsq = _interopRequireDefault(require("./../.stylex/ukchsq7"));
var _ubyhozq = _interopRequireDefault(require("./../.stylex/ubyhozq"));
var _uelbtaz = _interopRequireDefault(require("./../.stylex/uelbtaz"));
var _u99gbyh = _interopRequireDefault(require("./../.stylex/u99gbyh"));
var _u13d3rh = _interopRequireDefault(require("./../.stylex/u13d3rh8"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = require("@universe-design/stylex/react");
var _MetricsHeader = require("./MetricsHeader");
var _MetricsMainContent = require("./MetricsMainContent");
var _MetricsChart = require("./MetricsChart");
var _MetricsSecondaryContent = require("./MetricsSecondaryContent");
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _MetricsContext = require("./MetricsContext");
var _internal = require("../internal");
var React = _interopRequireWildcard(require("react"));
var _reactUtilities = require("@universe-design/react-utilities");
var _useMetrics = require("./useMetrics");
var _MetricsGroup = require("./MetricsGroup");
var _MetricsFooter = require("./MetricsFooter");
//.usbu5mz{gap:12px;}
//.u4xv25i{gap:4px;}
//.u1vx9l62{flex-direction:column;}
//.u5otk8j{border-top-left-radius:@border-radius-lg;}
//.u9kun4h{border-top-right-radius:@border-radius-lg;}
//.u1a4u8w2{border-bottom-right-radius:@border-radius-lg;}
//.uiy6t15{border-bottom-left-radius:@border-radius-lg;}
//.u22iagw{display:flex;}
//.u122n59{align-items:center;}
//.u19gb1f4{gap:8px;}
//.uqerorx{flex-grow:1;}
//.uly5x3f{width:100%;}
/* eslint-disable react-hooks/rules-of-hooks */
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const sxMetricsValue = {
  root: {
    "props": {
      "Bh6795r": _u99gbyh.default,
      "a9b677": _u13d3rh.default
    }
  }
};
const MetricsValue = (0, _internal.forwardHTMLRef)((props, ref) => (0, _react.jsx)("div", {
  ref: ref,
  className: (0, _stylex.default)(sxMetricsValue.root)
}, props.children));
const useMetricsStylexStyles = (0, _react.makeStylexStyles)(_ref => {
  let {
    prefix
  } = _ref;
  return {
    root: {
      "props": {
        "Btl43ni": _u1d70x8k.default,
        "B7oj6ja": _u1cpp6s.default,
        "Bbmb7ep": _uyretdy.default,
        "Beyfa6y": _unqnxb.default,
        "mc9l5x": _ukchsq.default,
        "Bt984gj": _ubyhozq.default,
        "rmohyg": _uelbtaz.default
      },
      "alias": `${prefix}__root`
    },
    main: {
      "props": {
        "rmohyg": _umue7z.default,
        "mc9l5x": _ukchsq.default,
        "Beiy3e4": _uohwdt.default
      },
      "alias": `${prefix}__main`
    },
    'secondary-root': {
      "props": {
        "mc9l5x": _ukchsq.default,
        "Beiy3e4": _uohwdt.default
      },
      "alias": `${prefix}__secondary-root`
    },
    'secondary-medium': {
      "props": {
        "rmohyg": _uelbtaz.default
      },
      "alias": `${prefix}__secondary-medium`
    },
    'secondary-large': {
      "props": {
        "rmohyg": _u40u48c.default
      },
      "alias": `${prefix}__secondary-large`
    }
  };
});
const MetricsRoot = exports.MetricsRoot = /*#__PURE__*/React.memo((0, _internal.forwardHTMLRef)((props, ref) => {
  const {
    mainTitle,
    mainInfoIcon,
    mainHeaderContentAfter,
    mainValue,
    secondary,
    footer,
    chart,
    ...restProps
  } = props;
  const {
    register
  } = (0, _useMetrics.useMetricsDescendant)();
  const {
    rootCssScope,
    size,
    position
  } = (0, _MetricsContext.useMetricsContext)();
  const recalculateLocation = (0, _MetricsGroup.useMetricsGroupContextSelector)(context => context === null || context === void 0 ? void 0 : context.recalculateLocation);
  const styles = useMetricsStylexStyles({
    prefix: rootCssScope()
  });
  return (0, _react.jsx)(_reactUtilities.ResizeObserver, {
    onResize: recalculateLocation
  }, (0, _react.jsx)("div", (0, _extends2.default)({
    ref: (0, _reactUtilities.useMergedRefs)(ref, register),
    sx: [styles.root]
  }, restProps), (0, _react.jsx)(MetricsValue, null, (0, _react.jsx)("div", {
    sx: [styles[`secondary-${size}`], styles['secondary-root']]
  }, (0, _react.jsx)("div", {
    sx: [styles.main]
  }, (0, _react.jsx)(_MetricsHeader.MetricsHeader, {
    mainTitle: mainTitle,
    mainInfoIcon: mainInfoIcon,
    mainHeaderContentAfter: mainHeaderContentAfter
  }), (0, _react.jsx)(_MetricsMainContent.MetricsMainContent, {
    mainValue: mainValue,
    chart: chart
  })), (0, _react.jsx)(_MetricsSecondaryContent.MetricsSecondaryContent, {
    secondary: secondary
  })), (0, _react.jsx)(_MetricsFooter.MetricsFooter, {
    footer: footer
  })), position === 'afterAll' && (0, _react.jsx)(_MetricsChart.MetricsChart, {
    chart: chart
  })));
}));
