// @universe-design/react/lib/metrics/MetricsMainContent.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MetricsMainContent = void 0;
var _u1u1p7mj = _interopRequireDefault(require("./../.stylex/u1u1p7mj"));
var _uc4lfdq = _interopRequireDefault(require("./../.stylex/uc4lfdq"));
var _uugkpzq = _interopRequireDefault(require("./../.stylex/uugkpzq"));
var _u5jxv = _interopRequireDefault(require("./../.stylex/u5jxv71"));
var _u10wmqsv = _interopRequireDefault(require("./../.stylex/u10wmqsv"));
var _u10if3fb = _interopRequireDefault(require("./../.stylex/u10if3fb"));
var _udavndu = _interopRequireDefault(require("./../.stylex/udavndu"));
var _u903f = _interopRequireDefault(require("./../.stylex/u903f20"));
var _u13d3rh = _interopRequireDefault(require("./../.stylex/u13d3rh8"));
var _ukchsq = _interopRequireDefault(require("./../.stylex/ukchsq7"));
var _ubyhozq = _interopRequireDefault(require("./../.stylex/ubyhozq"));
var _u125g0x = _interopRequireDefault(require("./../.stylex/u125g0x8"));
var _uwg391m = _interopRequireDefault(require("./../.stylex/uwg391m"));
var _umbknfa = _interopRequireDefault(require("./../.stylex/umbknfa"));
var _u22yt5q = _interopRequireDefault(require("./../.stylex/u22yt5q"));
var _u1tfkhub = _interopRequireDefault(require("./../.stylex/u1tfkhub"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = require("@universe-design/stylex/react");
var _reactUtilities = require("@universe-design/react-utilities");
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _react2 = _interopRequireWildcard(require("react"));
var _MetricsChart = require("./MetricsChart");
var _forwardRef = require("../util/forwardRef");
var _MetricsContext = require("./MetricsContext");
var _slot = require("../shared/slot");
var _useMetrics = require("./useMetrics");
var _MetricsGroup = require("./MetricsGroup");
//.u1euv43f{position:absolute;}
//.u1b6n930{transform-origin:left top;}
//.uz5stix{white-space:nowrap;}
//.u10pi13n{position:relative;}
//.u1jo26qt{font-size:36px;}
//.ut8aq9l{line-height:42px;}
//.u1rt2boy{font-size:24px;}
//.u336tjw{line-height:28px;}
//.uly5x3f{width:100%;}
//.u22iagw{display:flex;}
//.u122n59{align-items:center;}
//.ulvyvdh{flex-wrap:nowrap;}
//.u1869bpl{justify-content:space-between;}
//.ueyvcx7{font-family:Roboto;}
//.u1z30l8{color:@text-title;}
//.u1hd9hm2{font-weight:500;}
/* eslint-disable react/no-unknown-property */
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const useMetricsMainContentStylexStyles = (0, _react.makeStylexStyles)(_ref => {
  let {
    prefix
  } = _ref;
  return {
    root: {
      "props": {
        "a9b677": _u13d3rh.default,
        "mc9l5x": _ukchsq.default,
        "Bt984gj": _ubyhozq.default,
        "Eh141a": _u125g0x.default,
        "Brf1p80": _uwg391m.default,
        "Bahqtrf": _umbknfa.default,
        "sj55zd": _u22yt5q.default,
        "Bhrd7zp": _u1tfkhub.default
      },
      "alias": `${prefix}__root`
    },
    mainWrapper: {
      "props": {
        "mc9l5x": _ukchsq.default,
        "Bt984gj": _ubyhozq.default
      },
      "alias": `${prefix}__main__wrapper`
    },
    medium: {
      "props": {
        "Be2twd7": _udavndu.default,
        "Bhrd7zp": _u1tfkhub.default,
        "Bg96gwp": _u903f.default
      },
      "alias": `${prefix}__medium`
    },
    large: {
      "props": {
        "Be2twd7": _u10wmqsv.default,
        "Bg96gwp": _u10if3fb.default
      },
      "alias": `${prefix}__large`
    },
    container: {
      "props": {
        "a9b677": _u13d3rh.default,
        "qhf8xq": _u5jxv.default
      },
      "alias": `${prefix}__container`
    },
    content: {
      "props": {
        "qhf8xq": _u1u1p7mj.default,
        "B76gyc8": _uc4lfdq.default,
        "Huce71": _uugkpzq.default
      },
      "alias": `${prefix}__content`
    }
  };
});
const MetricsMainContent = exports.MetricsMainContent = (0, _forwardRef.forwardHTMLRef)((props, ref) => {
  const {
    mainValue,
    chart,
    ...restProps
  } = props;
  const {
    rootCssScope,
    size,
    position
  } = (0, _MetricsContext.useMetricsContext)();
  const {
    index,
    register
  } = (0, _useMetrics.useMetricsMainDescendant)();
  const recalculateMainHeights = (0, _MetricsGroup.useMetricsGroupContextSelector)(context => context === null || context === void 0 ? void 0 : context.recalculateMainHeights);
  const mainRowMaxHeight = (0, _MetricsGroup.useMetricsGroupContextSelector)(context => {
    var _context$mainRowInfo, _context$mainRowInfo$;
    return context === null || context === void 0 ? void 0 : (_context$mainRowInfo = context.mainRowInfo) === null || _context$mainRowInfo === void 0 ? void 0 : (_context$mainRowInfo$ = _context$mainRowInfo[index]) === null || _context$mainRowInfo$ === void 0 ? void 0 : _context$mainRowInfo$.maxHeight;
  });
  const styles = useMetricsMainContentStylexStyles({
    prefix: rootCssScope()
  });
  const [containerWidth, setContainerWidth] = _react2.default.useState(1);
  const [contentWidth, setContentWidth] = _react2.default.useState(1);
  const [contentHeight, setContentHeight] = _react2.default.useState(1);
  const scale = (0, _react2.useMemo)(() => {
    if (containerWidth < contentWidth) {
      return containerWidth / contentWidth;
    }
    return 1;
  }, [containerWidth, contentWidth]);
  const mainValueProps = (0, _slot.resolveShorthandOld)(mainValue);
  const {
    autoScaleContent = true
  } = mainValueProps;
  let content;
  if (!autoScaleContent) {
    content = (0, _react.jsx)("span", null, mainValueProps === null || mainValueProps === void 0 ? void 0 : mainValueProps.children);
  } else {
    content = (0, _react.jsx)(_reactUtilities.ResizeObserver, {
      onResize: _ref2 => {
        let {
          width,
          height
        } = _ref2;
        setContentWidth(width);
        setContentHeight(height);
      }
    }, (0, _react.jsx)("span", {
      sx: [styles.content],
      style: {
        transform: `scale(${scale})`
      }
    }, mainValueProps === null || mainValueProps === void 0 ? void 0 : mainValueProps.children));
  }
  return (0, _react.jsx)("div", (0, _extends2.default)({
    sx: [styles.mainWrapper],
    style: {
      height: mainRowMaxHeight
    }
  }, restProps), (0, _react.jsx)(_reactUtilities.ResizeObserver, {
    onResize: recalculateMainHeights
  }, (0, _react.jsx)("div", {
    ref: (0, _reactUtilities.useMergedRefs)(ref, register),
    sx: [styles.root, styles[size]]
  }, (0, _react.jsx)(_reactUtilities.ResizeObserver, {
    onResize: _ref3 => {
      let {
        width
      } = _ref3;
      setContainerWidth(width);
    }
  }, (0, _react.jsx)("div", {
    sx: [styles.container],
    style: autoScaleContent ? {
      height: contentHeight * scale
    } : {}
  }, content)), position === 'afterMainValue' && (0, _react.jsx)(_MetricsChart.MetricsChart, {
    chart: chart
  }))));
});
