// @universe-design/react/lib/metrics/MetricsGroup.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useMetricsGroupContextSelector = exports.MetricsGroup = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _reactProvider = require("@universe-design/react-provider");
var _useContextSelector = require("use-context-selector");
var React = _interopRequireWildcard(require("react"));
var _useMetrics = require("./useMetrics");
var _reactUtilities = require("@universe-design/react-utilities");
var _internal = require("../internal");
var _configProvider = require("../config-provider");
var _clsx = _interopRequireDefault(require("clsx"));
var _shim = require("use-sync-external-store/shim");
const {
  useContextSelector: useMetricsGroupContextSelector,
  Provider: MetricsGroupContextProvider
} = (0, _reactProvider.createContextHook)('MetricsGroupContext', {
  strict: false,
  customizedContextMethods: {
    createContext: _useContextSelector.createContext,
    useContext: _useContextSelector.useContext,
    useContextSelector: _useContextSelector.useContextSelector
  },
  defaultValue: null
});
exports.useMetricsGroupContextSelector = useMetricsGroupContextSelector;
const MetricsGroup = exports.MetricsGroup = (0, _internal.forwardHTMLRef)((props, ref) => {
  const {
    titleLineClamp = 2,
    className,
    ...restProps
  } = props;
  const [location, setLocation] = React.useState([]);
  const [headerHeights, setHeaderHeights] = React.useState([]);
  const [mainHeights, setMainHeights] = React.useState([]);
  const [secondaryHeights, setSecondaryHeights] = React.useState([]);
  const [footerHeights, setFooterHeights] = React.useState([]);
  const rootCssScope = (0, _configProvider.useCssScope)('metrics');
  const descendants = (0, _useMetrics.useMetricsDescendants)();
  const {
    list: descendantList
  } = (0, _shim.useSyncExternalStore)(descendants.subscribe, descendants.getSnapshot, descendants.getSnapshot);
  const headerDescendants = (0, _useMetrics.useMetricsHeaderDescendants)();
  const {
    list: headerDescendantList
  } = (0, _shim.useSyncExternalStore)(headerDescendants.subscribe, headerDescendants.getSnapshot, headerDescendants.getSnapshot);
  const mainDescendants = (0, _useMetrics.useMetricsMainDescendants)();
  const {
    list: mainDescendantList
  } = (0, _shim.useSyncExternalStore)(mainDescendants.subscribe, mainDescendants.getSnapshot, mainDescendants.getSnapshot);
  const secondaryDescendants = (0, _useMetrics.useMetricsSecondaryDescendants)();
  const {
    list: secondaryDescendantList
  } = (0, _shim.useSyncExternalStore)(secondaryDescendants.subscribe, secondaryDescendants.getSnapshot, secondaryDescendants.getSnapshot);
  const footerDescendants = (0, _useMetrics.useMetricsFooterDescendants)();
  const {
    list: footerDescendantList
  } = (0, _shim.useSyncExternalStore)(footerDescendants.subscribe, footerDescendants.getSnapshot, footerDescendants.getSnapshot);
  const recalculateLocation = React.useCallback(() => {
    const xy = [];
    for (const d of descendantList) {
      xy.push({
        x: (d === null || d === void 0 ? void 0 : d.node.offsetLeft) || 0,
        y: (d === null || d === void 0 ? void 0 : d.node.offsetTop) || 0
      });
    }
    setLocation(xy);
  }, [descendantList]);
  const recalculateHeaderHeights = React.useCallback(() => {
    const headerHeight = [];
    for (const d of headerDescendantList) {
      headerHeight.push(d === null || d === void 0 ? void 0 : d.node.clientHeight);
    }
    setHeaderHeights(headerHeight);
  }, [headerDescendantList]);
  const recalculateMainHeights = React.useCallback(() => {
    const mainHeight = [];
    for (const d of mainDescendantList) {
      mainHeight.push(d === null || d === void 0 ? void 0 : d.node.clientHeight);
    }
    setMainHeights(mainHeight);
  }, [mainDescendantList]);
  const recalculateSecondaryHeights = React.useCallback(() => {
    const secondaryHeight = [];
    for (const d of secondaryDescendantList) {
      secondaryHeight.push(d === null || d === void 0 ? void 0 : d.node.clientHeight);
    }
    setSecondaryHeights(secondaryHeight);
  }, [secondaryDescendantList]);
  const recalculateFooterHeights = React.useCallback(() => {
    const footerHeight = [];
    for (const d of footerDescendantList) {
      footerHeight.push(d === null || d === void 0 ? void 0 : d.node.clientHeight);
    }
    setFooterHeights(footerHeight);
  }, [footerDescendantList]);
  React.useEffect(() => {
    recalculateLocation();
    recalculateHeaderHeights();
    recalculateMainHeights();
    recalculateSecondaryHeights();
    recalculateFooterHeights();
  }, [recalculateLocation, recalculateHeaderHeights, recalculateMainHeights, recalculateSecondaryHeights, recalculateFooterHeights]);
  const calculateRowInfo = React.useCallback(heights => {
    const rows = [];

    // 按 y 坐标分组
    location.forEach((loc, index) => {
      const existingRow = rows.find(row => row.y === loc.y);
      if (existingRow) {
        existingRow.heights.push(heights[index]);
      } else {
        rows.push({
          y: loc.y,
          heights: [heights[index]]
        });
      }
    });

    // 计算每行的最大高度映射
    const rowMaxHeights = rows.map(row => ({
      y: row.y,
      maxHeight: Math.max(...row.heights)
    }));

    // 为每个位置生成对应的行信息
    return location.map(loc => {
      var _rowMaxHeights$find;
      return {
        row: rows.findIndex(r => r.y === loc.y),
        maxHeight: ((_rowMaxHeights$find = rowMaxHeights.find(r => r.y === loc.y)) === null || _rowMaxHeights$find === void 0 ? void 0 : _rowMaxHeights$find.maxHeight) || 0
      };
    });
  }, [location]);
  const headerRowInfo = React.useMemo(() => calculateRowInfo(headerHeights), [calculateRowInfo, headerHeights]);
  const mainRowInfo = React.useMemo(() => calculateRowInfo(mainHeights), [calculateRowInfo, mainHeights]);
  const secondaryRowInfo = React.useMemo(() => calculateRowInfo(secondaryHeights), [calculateRowInfo, secondaryHeights]);
  const footerRowInfo = React.useMemo(() => calculateRowInfo(footerHeights), [calculateRowInfo, footerHeights]);
  return /*#__PURE__*/React.createElement(_useMetrics.MetricsDescendantsProvider, {
    value: descendants
  }, /*#__PURE__*/React.createElement(_useMetrics.MetricsHeaderDescendantsProvider, {
    value: headerDescendants
  }, /*#__PURE__*/React.createElement(_useMetrics.MetricsSecondaryDescendantsProvider, {
    value: secondaryDescendants
  }, /*#__PURE__*/React.createElement(_useMetrics.MetricsFooterDescendantsProvider, {
    value: footerDescendants
  }, /*#__PURE__*/React.createElement(_useMetrics.MetricsMainDescendantsProvider, {
    value: mainDescendants
  }, /*#__PURE__*/React.createElement(MetricsGroupContextProvider, {
    value: {
      titleLineClamp,
      recalculateLocation,
      recalculateMainHeights,
      recalculateHeaderHeights,
      recalculateSecondaryHeights,
      recalculateFooterHeights,
      headerRowInfo,
      mainRowInfo,
      secondaryRowInfo,
      footerRowInfo
    }
  }, /*#__PURE__*/React.createElement(_reactUtilities.ResizeObserver, {
    onResize: recalculateLocation
  }, /*#__PURE__*/React.createElement("div", (0, _extends2.default)({
    className: (0, _clsx.default)(`${rootCssScope()}__group__wrapper`, className),
    ref: ref
  }, restProps), props.children))))))));
});
