// @universe-design/react/lib/metrics/MetricsSecondaryContent.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MetricsSecondaryContent = void 0;
var _u1xg5upn = _interopRequireDefault(require("./../.stylex/u1xg5upn"));
var _u1amwvao = _interopRequireDefault(require("./../.stylex/u1amwvao"));
var _ucy8jds = _interopRequireDefault(require("./../.stylex/ucy8jds"));
var _uv1n46n = _interopRequireDefault(require("./../.stylex/uv1n46n"));
var _u894z0b = _interopRequireDefault(require("./../.stylex/u894z0b"));
var _u1agd5ph = _interopRequireDefault(require("./../.stylex/u1agd5ph"));
var _u1e9he2z = _interopRequireDefault(require("./../.stylex/u1e9he2z"));
var _umbknfa = _interopRequireDefault(require("./../.stylex/umbknfa"));
var _ueh72r = _interopRequireDefault(require("./../.stylex/ueh72r3"));
var _u130iqk = _interopRequireDefault(require("./../.stylex/u130iqk6"));
var _u8wgnjj = _interopRequireDefault(require("./../.stylex/u8wgnjj"));
var _uugkpzq = _interopRequireDefault(require("./../.stylex/uugkpzq"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = require("@universe-design/stylex/react");
var _ExpandDownFilled = _interopRequireDefault(require("@universe-design/icons-react/ExpandDownFilled"));
var _ExpandUpFilled = _interopRequireDefault(require("@universe-design/icons-react/ExpandUpFilled"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _react2 = _interopRequireWildcard(require("react"));
var _divider = require("../divider");
var _text = require("../text");
var _reactUtilities = require("@universe-design/react-utilities");
var _forwardRef = require("../util/forwardRef");
var _MetricsContext = require("./MetricsContext");
var _constant = require("./constant");
var _useMetrics = require("./useMetrics");
var _MetricsGroup = require("./MetricsGroup");
var _clsx = _interopRequireDefault(require("clsx"));
//.u112evzf{color:@datav-function-element-neutral;}
//.u19ci1c7{color:@datav-function-element-warning;}
//.uy5pp46{color:@datav-function-element-info;}
//.u1bxglv9{color:@datav-function-element-danger;}
//.u1o67kls{color:@datav-function-element-success;}
//.u1hzsgrp{margin-left:2px;}
//.ulkuc6h{line-height:100%;}
//.ueyvcx7{font-family:Roboto;}
//.u1rvp7gu{color:@text-caption;}
//.u8zmjen{margin-right:4px;}
//.u14t3ns0{display:inline-block;}
//.uz5stix{white-space:nowrap;}
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
function isPresetColor(color) {
  return _constant.presetColor.includes(color);
}
const SecondaryItem = _ref => {
  let {
    index,
    setItemWidths,
    item
  } = _ref;
  return (0, _react.jsx)(_reactUtilities.ResizeObserver, {
    onResize: _ref2 => {
      let {
        width
      } = _ref2;
      setItemWidths(prev => {
        const newWidths = [...prev];
        newWidths[index] = width;
        return newWidths;
      });
    }
  }, (0, _react.jsx)(MetricsSecondaryContentItem, (0, _extends2.default)({
    ref: (0, _react2.useCallback)(node => {
      if (!node) {
        setItemWidths(prev => {
          const newWidths = [...prev];
          newWidths.splice(index, 1);
          return newWidths;
        });
      }
    }, [index, setItemWidths])
  }, item)));
};
const MetricsSecondaryContent = exports.MetricsSecondaryContent = (0, _forwardRef.forwardHTMLRef)((props, ref) => {
  const {
    secondary,
    className,
    ...restProps
  } = props;
  const secondaryInfo = secondary === null || secondary === void 0 ? void 0 : secondary.data;
  const [itemWidths, setItemWidths] = (0, _react2.useState)([]);
  const [containerWidth, setContainerWidth] = (0, _react2.useState)(0);
  const [dividerWidth, setDividerWidth] = (0, _react2.useState)(0);
  const {
    index,
    register
  } = (0, _useMetrics.useMetricsSecondaryDescendant)();
  const recalculateSecondaryHeights = (0, _MetricsGroup.useMetricsGroupContextSelector)(context => context === null || context === void 0 ? void 0 : context.recalculateSecondaryHeights);
  const secondaryRowMaxHeight = (0, _MetricsGroup.useMetricsGroupContextSelector)(context => {
    var _context$secondaryRow, _context$secondaryRow2;
    return context === null || context === void 0 ? void 0 : (_context$secondaryRow = context.secondaryRowInfo) === null || _context$secondaryRow === void 0 ? void 0 : (_context$secondaryRow2 = _context$secondaryRow[index]) === null || _context$secondaryRow2 === void 0 ? void 0 : _context$secondaryRow2.maxHeight;
  });
  const {
    rootCssScope
  } = (0, _MetricsContext.useMetricsContext)();
  const styles = useMetricsSecondaryItemStylexStyles({
    prefix: rootCssScope()
  });
  const layoutSingleRow = _react2.default.useCallback((items, dividerWidth, startIndex) => {
    const row = [];
    let totalWidth = itemWidths[startIndex];

    // µĘ╗ÕŖĀń¼¼õĖĆõĖ¬Õģāń┤Ā
    row.push({
      item: items[startIndex],
      isLast: false,
      marginRight: 0,
      width: itemWidths[startIndex]
    });
    let currentIndex = startIndex + 1;

    // Õ░ØĶ»ĢµĘ╗ÕŖĀµø┤ÕżÜÕģāń┤Ā
    while (currentIndex < items.length) {
      const nextItemWidth = dividerWidth + itemWidths[currentIndex];
      const newTotalWidth = totalWidth + nextItemWidth;
      if (newTotalWidth > containerWidth) {
        // ÕĮōÕēŹĶĪīµöŠõĖŹõĖŗµø┤ÕżÜÕģāń┤Ā
        const lastItem = row[row.length - 1];
        lastItem.isLast = true;
        lastItem.marginRight = Math.floor(containerWidth - totalWidth) || 0;
        break;
      }

      // ÕÅ»õ╗źń╗¦ń╗ŁµĘ╗ÕŖĀÕģāń┤Ā
      row.push({
        item: items[currentIndex],
        isLast: false,
        marginRight: 0,
        width: itemWidths[currentIndex]
      });
      totalWidth = newTotalWidth;
      currentIndex++;
    }

    // ÕżäńÉåµ£ĆÕÉÄõĖĆĶĪī
    if (currentIndex === items.length) {
      const lastItem = row[row.length - 1];
      lastItem.isLast = true;
      lastItem.marginRight = Math.floor(containerWidth - totalWidth) || 0;
    }
    return row;
  }, [containerWidth, itemWidths]);
  const secondaryList = _react2.default.useMemo(() => {
    if (!secondaryInfo) {
      return [];
    }
    const result = [];
    let currentIndex = 0;
    while (currentIndex < secondaryInfo.length) {
      const row = layoutSingleRow(secondaryInfo, dividerWidth, currentIndex);
      result.push(...row);
      currentIndex += row.length;
    }
    return result;
  }, [secondaryInfo, layoutSingleRow, dividerWidth]);
  if (!secondaryInfo) {
    return null;
  }
  return (0, _react.jsx)(_react2.default.Fragment, null, (0, _react.jsx)("div", (0, _extends2.default)({
    className: (0, _clsx.default)(`${rootCssScope()}__secondary__wrapper`, className),
    style: {
      height: secondaryRowMaxHeight
    }
  }, restProps), (0, _react.jsx)(_reactUtilities.ResizeObserver, {
    onResize: _ref3 => {
      let {
        width
      } = _ref3;
      setContainerWidth(width);
      recalculateSecondaryHeights === null || recalculateSecondaryHeights === void 0 ? void 0 : recalculateSecondaryHeights();
    }
  }, (0, _react.jsx)("div", {
    ref: (0, _reactUtilities.useMergedRefs)(ref, register)
  }, secondaryList.map((_ref4, i) => {
    let {
      item,
      isLast,
      marginRight
    } = _ref4;
    return (0, _react.jsx)(_react2.default.Fragment, {
      key: i
    }, (0, _react.jsx)(SecondaryItem, {
      index: i,
      item: item,
      setItemWidths: setItemWidths
    }), isLast && marginRight > 0 && (0, _react.jsx)("span", {
      style: {
        width: marginRight
      },
      sx: [styles.spacer]
    }), !isLast && (0, _react.jsx)(_reactUtilities.ResizeObserver, {
      onResize: _ref5 => {
        let {
          width
        } = _ref5;
        setDividerWidth(width);
      }
    }, (0, _react.jsx)("span", {
      style: {
        display: 'inline-block'
      }
    }, (0, _react.jsx)(_divider.Divider, {
      type: "vertical"
    }))));
  })))));
});
const useMetricsSecondaryItemStylexStyles = (0, _react.makeStylexStyles)(_ref6 => {
  let {
    prefix
  } = _ref6;
  return {
    root: {
      "props": {
        "mc9l5x": _u8wgnjj.default,
        "Huce71": _uugkpzq.default
      },
      "alias": `${prefix}__root`
    },
    spacer: {
      "props": {
        "mc9l5x": _u8wgnjj.default
      },
      "alias": `${prefix}__spacer`
    },
    label: {
      "props": {
        "sj55zd": _ueh72r.default,
        "t21cq0": _u130iqk.default
      },
      "alias": `${prefix}__label`
    },
    icon: {
      "props": {
        "Bahqtrf": _umbknfa.default
      },
      "alias": `${prefix}__icon`
    },
    value: {
      "props": {
        "Frg6f3": _u1agd5ph.default,
        "Bahqtrf": _umbknfa.default,
        "Bg96gwp": _u1e9he2z.default
      },
      "alias": `${prefix}__value`
    },
    success: {
      "props": {
        "sj55zd": _u894z0b.default
      },
      "alias": `${prefix}__success`
    },
    danger: {
      "props": {
        "sj55zd": _uv1n46n.default
      },
      "alias": `${prefix}__danger`
    },
    info: {
      "props": {
        "sj55zd": _ucy8jds.default
      },
      "alias": `${prefix}__info`
    },
    warning: {
      "props": {
        "sj55zd": _u1amwvao.default
      },
      "alias": `${prefix}__warning`
    },
    neutral: {
      "props": {
        "sj55zd": _u1xg5upn.default
      },
      "alias": `${prefix}__neutral`
    }
  };
});
const MetricsSecondaryContentItem = (0, _forwardRef.forwardHTMLRef)((props, ref) => {
  const {
    label,
    value: valueNode,
    type = 'none',
    valueStyle,
    color = 'neutral',
    ...restProps
  } = props;
  const {
    size,
    rootCssScope
  } = (0, _MetricsContext.useMetricsContext)();
  const iconMap = {
    increase: (0, _react.jsx)(_ExpandUpFilled.default, null),
    decrease: (0, _react.jsx)(_ExpandDownFilled.default, null),
    none: null
  };
  const iconNode = iconMap[type];
  const isPreset = isPresetColor(color);
  const styles = useMetricsSecondaryItemStylexStyles({
    prefix: rootCssScope()
  });
  return (0, _react.jsx)("span", (0, _extends2.default)({
    ref: ref,
    sx: [styles.root]
  }, restProps), label !== null && (0, _react.jsx)(_text.Text, {
    variant: "body0",
    sx: [styles.label]
  }, label), iconNode && (0, _react.jsx)(_text.Text, {
    sx: [styles.icon, isPreset ? styles[color] : null],
    style: {
      color: isPreset ? undefined : color,
      ...valueStyle
    },
    variant: "headline"
  }, iconNode), valueNode !== null && (0, _react.jsx)(_text.Text, {
    sx: [styles.value, isPreset ? styles[color] : null],
    style: {
      color: isPreset ? undefined : color,
      ...valueStyle
    },
    variant: size === 'medium' ? 'headline' : 'title4'
  }, valueNode));
});
