// @universe-design/react/lib/metrics/useMetrics.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useMetricsSecondaryDescendantsContext = exports.useMetricsSecondaryDescendants = exports.useMetricsSecondaryDescendant = exports.useMetricsMainDescendantsContext = exports.useMetricsMainDescendants = exports.useMetricsMainDescendant = exports.useMetricsHeaderDescendantsContext = exports.useMetricsHeaderDescendants = exports.useMetricsHeaderDescendant = exports.useMetricsFooterDescendantsContext = exports.useMetricsFooterDescendants = exports.useMetricsFooterDescendant = exports.useMetricsDescendantsContext = exports.useMetricsDescendants = exports.useMetricsDescendant = exports.MetricsSecondaryDescendantsProvider = exports.MetricsMainDescendantsProvider = exports.MetricsHeaderDescendantsProvider = exports.MetricsFooterDescendantsProvider = exports.MetricsDescendantsProvider = void 0;
var _reactUtilities = require("@universe-design/react-utilities");
// 注册 Metrics 最外面的 wrapper，用来获得组件的左上角坐标，判断是否在同一行
const [MetricsDescendantsProvider, useMetricsDescendantsContext, useMetricsDescendants, useMetricsDescendant] = (0, _reactUtilities.createDescendantContextOptional)();

// 注册标题区域
exports.useMetricsDescendant = useMetricsDescendant;
exports.useMetricsDescendants = useMetricsDescendants;
exports.useMetricsDescendantsContext = useMetricsDescendantsContext;
exports.MetricsDescendantsProvider = MetricsDescendantsProvider;
const [MetricsHeaderDescendantsProvider, useMetricsHeaderDescendantsContext, useMetricsHeaderDescendants, useMetricsHeaderDescendant] = (0, _reactUtilities.createDescendantContextOptional)();

// 注册主指标区域
exports.useMetricsHeaderDescendant = useMetricsHeaderDescendant;
exports.useMetricsHeaderDescendants = useMetricsHeaderDescendants;
exports.useMetricsHeaderDescendantsContext = useMetricsHeaderDescendantsContext;
exports.MetricsHeaderDescendantsProvider = MetricsHeaderDescendantsProvider;
const [MetricsMainDescendantsProvider, useMetricsMainDescendantsContext, useMetricsMainDescendants, useMetricsMainDescendant] = (0, _reactUtilities.createDescendantContextOptional)();

// 注册次级指标区域
exports.useMetricsMainDescendant = useMetricsMainDescendant;
exports.useMetricsMainDescendants = useMetricsMainDescendants;
exports.useMetricsMainDescendantsContext = useMetricsMainDescendantsContext;
exports.MetricsMainDescendantsProvider = MetricsMainDescendantsProvider;
const [MetricsSecondaryDescendantsProvider, useMetricsSecondaryDescendantsContext, useMetricsSecondaryDescendants, useMetricsSecondaryDescendant] = (0, _reactUtilities.createDescendantContextOptional)();

// 注册 Footer 区域
exports.useMetricsSecondaryDescendant = useMetricsSecondaryDescendant;
exports.useMetricsSecondaryDescendants = useMetricsSecondaryDescendants;
exports.useMetricsSecondaryDescendantsContext = useMetricsSecondaryDescendantsContext;
exports.MetricsSecondaryDescendantsProvider = MetricsSecondaryDescendantsProvider;
const [MetricsFooterDescendantsProvider, useMetricsFooterDescendantsContext, useMetricsFooterDescendants, useMetricsFooterDescendant] = (0, _reactUtilities.createDescendantContextOptional)();
exports.useMetricsFooterDescendant = useMetricsFooterDescendant;
exports.useMetricsFooterDescendants = useMetricsFooterDescendants;
exports.useMetricsFooterDescendantsContext = useMetricsFooterDescendantsContext;
exports.MetricsFooterDescendantsProvider = MetricsFooterDescendantsProvider;
