// @universe-design/react/src/divider/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import CHorizontal from '../demo/horizontal';
import CPlain from '../demo/plain';
import CVertical from '../demo/vertical';
import CWithText from '../demo/with-text';
if (!(CHorizontal as Story).storyName) {
  (CHorizontal as Story).storyName = '0. 基础用法';
  (CHorizontal as Story).parameters = {
    docs: {
      description: {
        story: '默认为水平分割线，可在中间加入文字。\n\n',
      },
    },
  };
}
if (!(CPlain as Story).storyName) {
  (CPlain as Story).storyName = '2. 分割文字使用正文样式';
  (CPlain as Story).parameters = {
    docs: {
      description: {
        story: '使用 `plain` 可以设置为更轻量的分割文字样式。\n\n',
      },
    },
  };
}
if (!(CVertical as Story).storyName) {
  (CVertical as Story).storyName = '3. 垂直分割线';
  (CVertical as Story).parameters = {
    docs: {
      description: {
        story: '使用 `type="vertical"` 设置为行内的垂直分割线。\n\n',
      },
    },
  };
}
if (!(CWithText as Story).storyName) {
  (CWithText as Story).storyName = '1. 带文字的分割线';
  (CWithText as Story).parameters = {
    docs: {
      description: {
        story: '分割线中带有文字，可以用 `orientation` 指定文字位置。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Divider/官网示例',
  tags: ['autodocs'],
};

export { CHorizontal, CPlain, CVertical, CWithText };
