// @universe-design/react/src/checkbox/stories/homeless.stories.tsx
/** @jsxImportSource @emotion/react */
import type { ComponentProps } from 'react';
import React, { useState } from 'react';
import { PerformanceContent } from '../../_stories/common/PerformanceContent';
import type { CheckboxOption } from '@universe-design/react';
import { Button, Checkbox, Row, Col, Switch, Space, Tooltip, Text, Divider, Select } from '@universe-design/react';

export default {
  title: 'Components/Checkbox/临时示例',
  component: Checkbox,
};

export const BasicUsage = Object.assign((props: ComponentProps<typeof Checkbox>) => <Checkbox {...props} />, {
  args: {
    label: '复选框',
  },
  argTypes: {
    label: { type: 'string' },
  },
});

export const DifferentStates = () => {
  const [disabled, setDisabled] = useState<boolean>(false);
  const [checkdAll, setCheckAll] = useState<boolean>();
  const [checkedList, setCheckedList] = useState<string[]>(['Apple']);
  const [indeterminate, setIndeterminate] = useState<boolean>(checkedList.length < 3 && checkedList.length > 0);
  const checkList: CheckboxOption[] = [
    { label: 'Apple', value: 'Apple' },
    { label: 'Pear', value: 'Pear' },
    { label: 'Orange', value: 'Orange', disabled },
  ];

  const checkListStr = ['Apple', 'Pear', 'Orange'];
  return (
    <div>
      <h2>CheckBox</h2>
      <div>
        <Checkbox aria-label="未选中" onChange={e => console.log(e)}>
          未选中
        </Checkbox>
        <Checkbox checked={false} aria-label="未选中">
          未选中
        </Checkbox>
        <Checkbox checked={false} aria-label="未选中" disabled>
          未选中
        </Checkbox>
        <Checkbox checked={false} aria-label="未选中" error>
          未选中
        </Checkbox>
        <Checkbox checked={true} aria-label="已选中">
          已选中
        </Checkbox>
        <Checkbox checked={true} aria-label="已选中" disabled>
          已选中
        </Checkbox>
        <Checkbox checked={false} indeterminate={true} aria-label="部分选中">
          部分选中
        </Checkbox>
      </div>
      <h2>CheckBox Group</h2>
      <Checkbox.Group onChange={values => console.log(values)}>
        <Checkbox value="Apple">Apple</Checkbox>
        <Checkbox value="Pear">Pear</Checkbox>
        <Checkbox value="Orange">Orange</Checkbox>
      </Checkbox.Group>
      <Divider text="默认选中" />
      <Checkbox.Group defaultValue={['Apple']} onChange={values => console.log(values)}>
        <Checkbox value="Apple">Apple</Checkbox>
        <Checkbox value="Pear">Pear</Checkbox>
        <Checkbox value="Orange">Orange</Checkbox>
      </Checkbox.Group>
      <Divider text="禁用" />
      <Button style={{ marginBottom: 10 }} onClick={() => setDisabled(!disabled)}>
        切换禁用状态
      </Button>
      <Checkbox.Group disabled={disabled} defaultValue={['Apple']} onChange={values => console.log(values)}>
        <Checkbox value="Apple">Apple</Checkbox>
        <Checkbox value="Pear">Pear</Checkbox>
        <Checkbox value="Orange" disabled>
          Orange
        </Checkbox>
      </Checkbox.Group>
      <Divider text="全选功能" />
      <Checkbox
        value="Apple"
        indeterminate={indeterminate}
        disabled={disabled}
        onChange={e => {
          if (e.target.checked) {
            setCheckedList(['Apple', 'Pear', 'Orange']);
          } else {
            setCheckedList([]);
          }
          setCheckAll(e.target.checked);
          setIndeterminate(false);
        }}
        checked={checkdAll}
      >
        全选
      </Checkbox>
      <Checkbox.Group
        value={checkedList}
        disabled={disabled}
        onChange={values => {
          const result = values as string[];
          setIndeterminate(result.length < 3 && result.length > 0);
          setCheckAll(result.length === 3);
          setCheckedList(result);

          console.log('checked list', result);
        }}
      >
        <Checkbox value="Apple">Apple</Checkbox>
        <Checkbox value="Pear">Pear</Checkbox>
        <Checkbox value="Orange">Orange</Checkbox>
      </Checkbox.Group>

      <Divider text="使用 option 完成快速配置" />
      <Checkbox.Group
        defaultValue={['Apple']}
        onChange={values => console.log('checked list', values)}
        options={checkList}
      />
      <Divider text="使用 option 完成快速配置 - option 支持 string 数组 options=['Apple', 'Pear', 'Orange']" />
      <Checkbox.Group
        defaultValue={['Apple']}
        onChange={values => console.log('checked list', values)}
        options={checkListStr}
      />
      <Divider text="与栅格组合使用" />
      <Checkbox.Group defaultValue={['Apple']} onChange={values => console.log(values)}>
        <Row>
          <Col span={8}>
            <Checkbox value="Apple">Apple</Checkbox>
          </Col>
          <Col span={8}>
            <Checkbox value="Pear">Pear</Checkbox>
          </Col>
          <Col span={8}>
            <Checkbox value="Orange" disabled>
              Orange
            </Checkbox>
          </Col>
        </Row>
        <Row>
          <Col span={8}>
            <Checkbox value="Apple2">Apple2</Checkbox>
          </Col>
          <Col span={8}>
            <Checkbox value="Pear2">Pear2</Checkbox>
          </Col>
          <Col span={8}>
            <Checkbox value="Orange2" disabled>
              Orange2
            </Checkbox>
          </Col>
        </Row>
      </Checkbox.Group>
    </div>
  );
};

export const SupportNumber = () => (
  <Checkbox.Group
    onChange={values => {
      console.log('checked list', values);
    }}
  >
    <Checkbox value={0}>Apple</Checkbox>
    <Checkbox value={1}>Pear</Checkbox>
    <Checkbox value={2}>Orange</Checkbox>
    <Checkbox value={3}>Orange</Checkbox>
  </Checkbox.Group>
);

export const Option = () => <Checkbox.Group options={['a', 'b', 'c']} />;

export const Error = () => {
  const [error, setError] = useState<boolean>(false);
  const [disabled, setDisabled] = useState<boolean>(false);
  return (
    <Space style={{ width: '100%' }} direction="vertical" size="large">
      <Space align="center">
        <span>错误：</span>
        <Switch checked={error} onChange={setError} />
      </Space>
      <Space align="center">
        <span>禁用：</span>
        <Switch checked={disabled} onChange={setDisabled} />
      </Space>
      <Checkbox error={error} disabled={disabled} />
      <Checkbox.Group error={error} disabled={disabled}>
        <Checkbox value={0}>Apple</Checkbox>
        <Checkbox value={1}>Pear</Checkbox>
        <Checkbox value={2}>Orange</Checkbox>
        <Checkbox value={3}>Orange</Checkbox>
      </Checkbox.Group>
    </Space>
  );
};

export const WithToolTip = () => (
  <Tooltip title="title message">
    <Checkbox>Apple</Checkbox>
  </Tooltip>
);

export const TransitionBlink = () => {
  const [checked, setChecked] = useState(false);
  return (
    <>
      <Tooltip title="title message">
        <Checkbox checked={checked} onChange={e => setChecked(e.target.checked)}>
          Apple
        </Checkbox>
      </Tooltip>
      <PerformanceContent componentName="input" count={1000} itemRenderer={() => <Select />} />
    </>
  );
};

export const MultiLineGroup = () => (
  <Checkbox.Group defaultValue={['Orange']}>
    <Checkbox value="Apple">
      <Text variant="plain" as="div" lineClamp={2} style={{ maxWidth: 180 }}>
        Apple Apple Apple Apple Apple Apple Apple Apple Apple Apple
      </Text>
    </Checkbox>
    <Checkbox value="Pear">Pear</Checkbox>
    <Checkbox value="Orange">Orange</Checkbox>
  </Checkbox.Group>
);
