// @universe-design/react/src/checkbox/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C7VerticalAlign from '../demo/7.vertical-align';
import CBasic from '../demo/basic';
import CCheckAll from '../demo/check-all';
import CControlled from '../demo/controlled';
import CDisabled from '../demo/disabled';
import CGroup from '../demo/group';
import CLayout from '../demo/layout';
import COption from '../demo/option';
if (!(C7VerticalAlign as Story).storyName) {
  (C7VerticalAlign as Story).storyName = '7. 垂直对齐';
  (C7VerticalAlign as Story).parameters = {
    docs: {
      description: {
        story: '勾选框文本内容有两种对齐方式，默认采用首行垂直对齐 `align-items: baseline;`。\n\n',
      },
    },
  };
}
if (!(CBasic as Story).storyName) {
  (CBasic as Story).storyName = '0. 基础用法';
  (CBasic as Story).parameters = {
    docs: {
      description: {
        story: '简单的用法。\n\n',
      },
    },
  };
}
if (!(CCheckAll as Story).storyName) {
  (CCheckAll as Story).storyName = '3. 全选';
  (CCheckAll as Story).parameters = {
    docs: {
      description: {
        story: '通过`indeterminate`实现半选效果。\n\n',
      },
    },
  };
}
if (!(CControlled as Story).storyName) {
  (CControlled as Story).storyName = '2. 受控模式';
  (CControlled as Story).parameters = {
    docs: {
      description: {
        story: '通过`checked`控制是否选中。\n\n',
      },
    },
  };
}
if (!(CDisabled as Story).storyName) {
  (CDisabled as Story).storyName = '1. 禁用状态';
  (CDisabled as Story).parameters = {
    docs: {
      description: {
        story: '禁用操作。\n\n',
      },
    },
  };
}
if (!(CGroup as Story).storyName) {
  (CGroup as Story).storyName = '4. 复选框组';
  (CGroup as Story).parameters = {
    docs: {
      description: {
        story: '一组复选框。\n\n',
      },
    },
  };
}
if (!(CLayout as Story).storyName) {
  (CLayout as Story).storyName = '6. 布局';
  (CLayout as Story).parameters = {
    docs: {
      description: {
        story: '配合栅格或其他布局组件实现灵活的布局。\n\n',
      },
    },
  };
}
if (!(COption as Story).storyName) {
  (COption as Story).storyName = '5. 复选框组 - 配置方式';
  (COption as Story).parameters = {
    docs: {
      description: {
        story: '通过配置`options`来渲染复选框。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Checkbox/官网示例',
  tags: ['autodocs'],
};

export { C7VerticalAlign, CBasic, CCheckAll, CControlled, CDisabled, CGroup, CLayout, COption };
