// @universe-design/react/src/switch/Switch.tsx
import clsx from 'clsx';
import React from 'react';
import omit from 'lodash/omit';
import { useControlledState } from '@universe-design/react-utilities';
import type { DefaultExtraSizeType, DefaultSizeType } from '../util';
import { getSizeAbbr, devWarning, warning } from '../util';
import { useComponentSize, useCssScope } from '../config-provider/hooks';
import { SpinLoadingSvg } from '../_assets/svgs';
import { forwardHTMLRef } from '../util/forwardRef';

import './style';

export type SwitchChangeEventHandler = (
  checked: boolean,
  event: React.MouseEvent<HTMLButtonElement, MouseEvent>,
) => void;

export interface SwitchExoticProps {
  prefixCls?: string;
  /**
   * 尺寸
   * @defaultValue `medium`
   */
  size?: Extract<DefaultSizeType | DefaultExtraSizeType, 'small' | 'medium'>;
  /**
   * `className` 属性透传
   */
  className?: string;
  /**
   * 打开状态的受控值
   */
  checked?: boolean;
  /**
   * 默认打开状态
   * @defaultValue false
   */
  defaultChecked?: boolean;
  /**
   * 选中状态变化的回调
   */
  onChange?: SwitchChangeEventHandler;
  /**
   * 点击开关的回调
   */
  onClick?: SwitchChangeEventHandler;
  /**
   * @deprecated UD 不建议出现选中标识
   */
  checkedChildren?: React.ReactNode;
  /**
   * @deprecated UD 不建议出现未选中标识
   */
  unCheckedChildren?: React.ReactNode;
  /**
   * 不可用状态
   * @defaultValue false
   */
  disabled?: boolean;
  /**
   * 加载状态。加载状态下，同时会叠加 disabled 态
   * @defaultValue false
   */
  loading?: boolean;
  /**
   * `autoFocus` 属性透传
   */
  autoFocus?: boolean;
  /**
   * `style` 属性透传
   */
  style?: React.CSSProperties;
  /**
   * `id` 属性透传
   */
  id?: string;
  /**
   * `title` 属性透传
   */
  title?: string;
}

export const Switch = /* #__PURE__*/ Object.assign(
  forwardHTMLRef<SwitchExoticProps, 'button'>(function (props, ref) {
    const ctxSize = useComponentSize();
    const {
      checked: checkedProp,
      defaultChecked = false,
      loading = false,
      disabled: disabledProp = false,
      size = ctxSize === 'large' ? 'medium' : ctxSize,
      onChange,
      onClick,
      className,
      style,
      id,
      title,
      autoFocus,
      unCheckedChildren,
      checkedChildren,
      prefixCls,
      ...restProp
    } = props;

    warning(
      !unCheckedChildren && !checkedChildren,
      devWarning(Switch.displayName || Switch.name, 'unCheckedChildren and checkedChildren are not support.'),
    );

    warning(
      'checked' in props || !('value' in props),
      devWarning('Switch', '`value` is not a valid prop, do you mean `checked`?'),
    );

    const cssScope = useCssScope('switch', prefixCls);
    const [checked, setChecked] = useControlledState(checkedProp ?? defaultChecked, checkedProp);
    const disabled = disabledProp || loading;
    const usedSize = getSizeAbbr(size);
    return (
      <button
        // form 会传给 switch error prop，switch 直接忽略掉
        {...omit(restProp, 'error')}
        onClick={React.useCallback(
          e => {
            if (!disabled) {
              setChecked(
                _checked => !_checked,
                _checked => {
                  onChange?.(_checked, e);
                  onClick?.(_checked, e);
                },
              );
            }
          },
          [disabled, onChange, onClick, setChecked],
        )}
        className={clsx(cssScope(), checked && cssScope('-checked'), disabled && cssScope('-disabled'), className, {
          [cssScope(`-${usedSize}`)]: usedSize,
        })}
        style={style}
        id={id}
        type="button"
        role="switch"
        autoFocus={autoFocus}
        title={title}
        ref={ref}
        data-testid="switch-button"
        aria-checked={checked}
      >
        <div className={cssScope('__handler')}>
          {loading ? <SpinLoadingSvg className={cssScope('__loading')} /> : null}
        </div>
      </button>
    );
  }),
  {
    __UD_SWITCH: true,
  },
);

/**
 * @docgen
 */
export type SwitchProps = React.ComponentPropsWithoutRef<typeof Switch>;

if (__DEV__) {
  Switch.displayName = 'Switch';
}
