// @universe-design/react/lib/collapse/CollapsePanel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CollapsePanel = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _noop = _interopRequireDefault(require("lodash/noop"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _PanelContent = require("./PanelContent");
var _DownBoldOutlined = _interopRequireDefault(require("@universe-design/icons-react/DownBoldOutlined"));
var _ExpandRightFilled = _interopRequireDefault(require("@universe-design/icons-react/ExpandRightFilled"));
var _reactTransitionGroup = require("react-transition-group");
var _transtion = require("./transtion");
var _button = require("../button");
var _stylex2 = require("../stylex");
require("./style");
/**
 *
 * @public
 */

function defaultExpandIcon(props, customIcon) {
  const {
    expandIconPosition,
    isActive,
    prefixCls,
    collapsible
  } = props;
  let icon = null;
  let size;
  if (expandIconPosition === 'left') {
    icon = customIcon || /*#__PURE__*/_react.default.createElement(_ExpandRightFilled.default, {
      style: {
        transform: isActive ? 'rotate(90deg)' : ''
      }
    });
    size = 10;
  } else {
    icon = customIcon || /*#__PURE__*/_react.default.createElement(_DownBoldOutlined.default, {
      style: {
        transform: isActive ? 'rotate(180deg)' : ''
      }
    });
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    size = 16;
  }
  return /*#__PURE__*/_react.default.createElement(_button.IconButton, {
    className: (0, _clsx.default)(`${prefixCls}-expand-icon`, {
      [`${prefixCls}-expand-icon--noaction`]: collapsible === 'header',
      [`${prefixCls}-expand-icon-unfold`]: isActive
    }),
    icon: icon,
    size: size
  });
}

/**
 * Collapse.Panel 的属性
 * @public
 */

const CollapsePanel = props => {
  const {
    collapsible,
    children,
    isActive,
    panelKey,
    onItemClick = _noop.default,
    header,
    extra,
    expandIcon,
    showArrow = true,
    forceRender,
    destroyInactivePanel,
    expandIconPosition,
    prefixCls,
    style,
    className,
    accordion
  } = props;
  const disabled = collapsible === 'disabled';
  const actualActive = !disabled && isActive;
  const handleClick = () => {
    if (disabled) {
      return;
    }
    onItemClick(String(panelKey));
  };
  return /*#__PURE__*/_react.default.createElement("div", {
    style: style,
    className: (0, _clsx.default)(`${prefixCls}-item`, {
      [`${prefixCls}-item-disabled`]: disabled,
      [`${prefixCls}-item-no-arrow`]: !showArrow,
      [`${prefixCls}-item-active`]: actualActive
    }, className)
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: `${prefixCls}-header`,
    onClick: () => {
      collapsible !== 'header' && handleClick();
    },
    role: accordion ? 'tab' : 'button',
    "aria-expanded": actualActive
  }, (() => {
    if (showArrow) {
      const panelProps = {
        isActive: actualActive,
        collapsible,
        key: panelKey,
        expandIconPosition,
        prefixCls
      };
      const icon = defaultExpandIcon(panelProps, expandIcon && expandIcon(panelProps));
      if (! /*#__PURE__*/_react.default.isValidElement(icon)) {
        return icon;
      }
      return /*#__PURE__*/_react.default.cloneElement(icon, {
        className: (0, _clsx.default)(icon.props.className, `${prefixCls}-arrow`)
      });
    }
    return null;
  })(), /*#__PURE__*/_react.default.createElement("span", {
    className: (0, _clsx.default)(`${prefixCls}-header-text`, (0, _stylex.default)(_stylex2.cssTypographyHeadline)),
    onClick: () => collapsible === 'header' && handleClick()
  }, header), extra && /*#__PURE__*/_react.default.createElement("div", {
    className: `${prefixCls}-extra`
  }, extra)), /*#__PURE__*/_react.default.createElement(_reactTransitionGroup.CSSTransition, (0, _extends2.default)({}, _transtion.collapseTransitionProps, {
    in: actualActive,
    mountOnEnter: !forceRender,
    unmountOnExit: destroyInactivePanel && !forceRender
  }), state => /*#__PURE__*/_react.default.createElement(_PanelContent.PanelContent, {
    isActive: actualActive,
    forceRender: forceRender,
    prefixCls: prefixCls,
    className: (0, _clsx.default)({
      [`${prefixCls}-content-hidden`]: state === 'exited' && !actualActive,
      [`${prefixCls}-content-transition`]: state === 'entering' || state === 'exiting'
    }),
    role: accordion ? 'tabpanel' : undefined
  }, children)));
};
exports.CollapsePanel = CollapsePanel;
