// @universe-design/react/lib/collapse/Collapse.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Collapse = void 0;
var _react = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _hooks = require("../config-provider/hooks");
var _util = require("../util");
var _noop = _interopRequireDefault(require("lodash/noop"));
require("./style");
function getActiveKeysArray(activeKey) {
  if (!activeKey) {
    return [];
  }
  let currentActiveKey = activeKey;
  if (!Array.isArray(currentActiveKey)) {
    currentActiveKey = currentActiveKey ? [currentActiveKey] : [];
  }
  return currentActiveKey.map(key => String(key));
}

/**
 * @public
 */

const Collapse = props => {
  const {
    prefixCls: prefixClsInProps,
    expandIconPosition = 'right',
    className,
    children,
    collapsible,
    accordion,
    activeKey: activeKeyProps,
    defaultActiveKey,
    destroyInactivePanel,
    onChange = _noop.default,
    expandIcon,
    style,
    type = 'default'
  } = props;
  const [activeKey, setActiveKey] = (0, _react.useState)(getActiveKeysArray(activeKeyProps !== null && activeKeyProps !== void 0 ? activeKeyProps : defaultActiveKey));
  const cssScope = (0, _hooks.useCssScope)('collapse');
  const prefixCls = prefixClsInProps || cssScope();
  const collapseClassName = (0, _clsx.default)(prefixCls, [`${prefixCls}-icon-position-${expandIconPosition}`], className, {
    [`${prefixCls}-obvious-title`]: type === 'obvious'
  });
  (0, _react.useEffect)(() => {
    if (activeKeyProps) {
      setActiveKey(getActiveKeysArray(activeKeyProps));
    }
  }, [activeKeyProps]);
  const updateActiveKey = newActiveKey => {
    if (!('activeKey' in props)) {
      setActiveKey(newActiveKey);
    }
    onChange(accordion ? newActiveKey[0] : newActiveKey);
  };
  const handleItemClick = panelKey => {
    let newActiveKey = [];
    if (accordion) {
      newActiveKey = activeKey[0] === panelKey ? [] : [panelKey];
    } else {
      const isActive = activeKey.indexOf(panelKey) > -1;
      if (isActive) {
        newActiveKey = activeKey.filter(key => key !== panelKey);
      } else {
        newActiveKey = activeKey.concat(panelKey);
      }
    }
    updateActiveKey(newActiveKey);
  };
  return /*#__PURE__*/_react.default.createElement("div", {
    style: style,
    className: collapseClassName,
    role: accordion ? 'tablist' : undefined
  }, (() => (0, _util.toFlatArray)(children).map((child, index) => {
    if (! /*#__PURE__*/(0, _react.isValidElement)(child) || typeof child.type === 'string') {
      return child;
    }
    const {
      collapsible: childCollapsible,
      destroyInactivePanel: childDestroyInactivePanel
    } = child.props;
    const mergeCollapsible = childCollapsible !== null && childCollapsible !== void 0 ? childCollapsible : collapsible;
    const mergeDestroyInactivePanel = childDestroyInactivePanel !== null && childDestroyInactivePanel !== void 0 ? childDestroyInactivePanel : destroyInactivePanel;
    const key = child.key || String(index);
    let isActive = false;
    if (accordion) {
      isActive = activeKey[0] === key;
    } else {
      isActive = activeKey.indexOf(key) > -1;
    }
    const childProps = {
      ...child.props,
      key,
      panelKey: key,
      collapsible: mergeCollapsible,
      isActive,
      onItemClick: handleItemClick,
      destroyInactivePanel: mergeDestroyInactivePanel,
      expandIconPosition,
      prefixCls,
      expandIcon,
      accordion
    };
    return /*#__PURE__*/_react.default.cloneElement(child, childProps);
  }))());
};
exports.Collapse = Collapse;
