// @universe-design/react/src/loading/index.tsx
import type { CSSProperties, ReactElement } from 'react';
import React from 'react';
import clsx from 'clsx';
import { useCssScope } from '../config-provider/hooks';
import type { DefaultExtraSizeType, DefaultSizeType } from '../util';
import { getSizeAbbr, isFunction } from '../util';
import { useLoading } from './useLoading';
import { SpinLoadingSvg } from '../_assets/svgs';

import './style';

export interface LoadingContextType {
  loading: boolean;
}

export interface LoadingProps {
  /**
   * `className` 属性透传
   */
  className?: string;
  /**
   * 是否显示加载组件
   * @defaultValue true
   */
  spinning?: boolean;
  /**
   * `style` 属性透传
   */
  style?: CSSProperties;
  /**
   * 尺寸大小
   * @defaultValue `medium`
   */
  size?: Extract<DefaultSizeType | DefaultExtraSizeType, 'small' | 'medium' | 'large'>;
  /**
   * 描述文案
   */
  tip?: string;
  /**
   * 是否延迟显示加载状态
   */
  delay?: number;
  /**
   * 当加载组件作为包裹元素的时候，会给子节点外面加一层包裹元素，这个属性就是包裹元素的 `className` 属性透传
   */
  wrapperClassName?: string;
  /**
   * 自定义加载指示符
   */
  indicator?: ReactElement;
  /**
   * 加载组件描述文案的布局方向
   * @defaultValue `vertical`
   */
  direction?: 'vertical' | 'horizontal';
  /**
   * 可以使用 render prop 模式获取组件是否处于 `loading` 状态
   */
  children?: ((ctx: LoadingContextType) => React.ReactNode) | React.ReactNode;
}

/**
 * 跟 LoadingProps 相同，兼容之前声明
 * @public
 */
export type SpinProps = LoadingProps;

export const Loading: React.FC<LoadingProps> = function Loading(props) {
  const {
    className,
    style,
    tip,
    size = 'medium',
    direction = 'vertical',
    indicator,
    children,
    wrapperClassName,
    spinning: spinningProp = true,
    delay,
  } = props;
  const cssScope = useCssScope('loading');
  const [spinning] = useLoading({ loading: spinningProp, delay });
  const usedSize = getSizeAbbr(size);
  const classNames = clsx({
    [cssScope()]: true,
    [cssScope(`-size-${usedSize}`)]: usedSize,
    [cssScope(`-direction-${direction}`)]: true,
    [`${className}`]: className,
  });

  const renderIndicator = () => {
    if (indicator) {
      return <div className={cssScope('__spinner')}>{indicator}</div>;
    }
    return (
      <div className={cssScope('__lottie')}>
        <SpinLoadingSvg />
      </div>
    );
  };

  const spinNode = spinning ? (
    <div className={classNames} style={style}>
      {renderIndicator()}
      {tip ? <div className={cssScope('__tip')}>{tip}</div> : null}
    </div>
  ) : null;

  if (children) {
    const containerClassName = clsx(cssScope('__container'), spinning && cssScope('__container-blur'));
    return (
      <div className={clsx(cssScope('-nested-loading'), wrapperClassName)}>
        {spinNode}
        <div className={containerClassName}>{isFunction(children) ? children({ loading: spinning }) : children}</div>
      </div>
    );
  }

  return spinNode;
};
