// @byted-service/bytedtrace/lib/gen/bytedtrace_pb.js
// source: bytedtrace.proto
/**
 * @fileoverview
 * @enhanceable
 * @suppress {missingRequire} reports error on implicit type usages.
 * @suppress {messageConventions} JS Compiler reports an error if a variable or
 *     field starts with 'MSG_' and isn't a translatable message.
 * @public
 */
// GENERATED CODE -- DO NOT EDIT!
/* eslint-disable */
// @ts-nocheck

var jspb = require('google-protobuf');
var goog = jspb;
var global = Function('return this')();

goog.exportSymbol('proto.ClientSpanProto', null, global);
goog.exportSymbol('proto.CustomSpanProto', null, global);
goog.exportSymbol('proto.EventProto', null, global);
goog.exportSymbol('proto.MetricProto', null, global);
goog.exportSymbol('proto.ServerSpanProto', null, global);
goog.exportSymbol('proto.SpanRef', null, global);
goog.exportSymbol('proto.Tag', null, global);
goog.exportSymbol('proto.TagType', null, global);
goog.exportSymbol('proto.TraceID', null, global);
goog.exportSymbol('proto.TransactionProto', null, global);
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.TransactionProto = function (opt_data) {
    jspb.Message.initialize(
        this,
        opt_data,
        0,
        -1,
        proto.TransactionProto.repeatedFields_,
        null
    );
};
goog.inherits(proto.TransactionProto, jspb.Message);
if (goog.DEBUG && !COMPILED) {
    /**
     * @public
     * @override
     */
    proto.TransactionProto.displayName = 'proto.TransactionProto';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.CustomSpanProto = function (opt_data) {
    jspb.Message.initialize(
        this,
        opt_data,
        0,
        -1,
        proto.CustomSpanProto.repeatedFields_,
        null
    );
};
goog.inherits(proto.CustomSpanProto, jspb.Message);
if (goog.DEBUG && !COMPILED) {
    /**
     * @public
     * @override
     */
    proto.CustomSpanProto.displayName = 'proto.CustomSpanProto';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.ServerSpanProto = function (opt_data) {
    jspb.Message.initialize(
        this,
        opt_data,
        0,
        -1,
        proto.ServerSpanProto.repeatedFields_,
        null
    );
};
goog.inherits(proto.ServerSpanProto, jspb.Message);
if (goog.DEBUG && !COMPILED) {
    /**
     * @public
     * @override
     */
    proto.ServerSpanProto.displayName = 'proto.ServerSpanProto';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.ClientSpanProto = function (opt_data) {
    jspb.Message.initialize(
        this,
        opt_data,
        0,
        -1,
        proto.ClientSpanProto.repeatedFields_,
        null
    );
};
goog.inherits(proto.ClientSpanProto, jspb.Message);
if (goog.DEBUG && !COMPILED) {
    /**
     * @public
     * @override
     */
    proto.ClientSpanProto.displayName = 'proto.ClientSpanProto';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.EventProto = function (opt_data) {
    jspb.Message.initialize(
        this,
        opt_data,
        0,
        -1,
        proto.EventProto.repeatedFields_,
        null
    );
};
goog.inherits(proto.EventProto, jspb.Message);
if (goog.DEBUG && !COMPILED) {
    /**
     * @public
     * @override
     */
    proto.EventProto.displayName = 'proto.EventProto';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.MetricProto = function (opt_data) {
    jspb.Message.initialize(
        this,
        opt_data,
        0,
        -1,
        proto.MetricProto.repeatedFields_,
        null
    );
};
goog.inherits(proto.MetricProto, jspb.Message);
if (goog.DEBUG && !COMPILED) {
    /**
     * @public
     * @override
     */
    proto.MetricProto.displayName = 'proto.MetricProto';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.Tag = function (opt_data) {
    jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.Tag, jspb.Message);
if (goog.DEBUG && !COMPILED) {
    /**
     * @public
     * @override
     */
    proto.Tag.displayName = 'proto.Tag';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.TraceID = function (opt_data) {
    jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.TraceID, jspb.Message);
if (goog.DEBUG && !COMPILED) {
    /**
     * @public
     * @override
     */
    proto.TraceID.displayName = 'proto.TraceID';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.SpanRef = function (opt_data) {
    jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.SpanRef, jspb.Message);
if (goog.DEBUG && !COMPILED) {
    /**
     * @public
     * @override
     */
    proto.SpanRef.displayName = 'proto.SpanRef';
}

/**
 * List of repeated fields within this message type.
 * @private {!Array<number>}
 * @const
 */
proto.TransactionProto.repeatedFields_ = [10, 11, 12, 23];

if (jspb.Message.GENERATE_TO_OBJECT) {
    /**
     * Creates an object representation of this proto.
     * Field names that are reserved in JavaScript and will be renamed to pb_name.
     * Optional fields that are not set will be set to undefined.
     * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
     * For the list of reserved names please see:
     *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
     * @param {boolean=} opt_includeInstance Deprecated. whether to include the
     *     JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @return {!Object}
     */
    proto.TransactionProto.prototype.toObject = function (opt_includeInstance) {
        return proto.TransactionProto.toObject(opt_includeInstance, this);
    };

    /**
     * Static version of the {@see toObject} method.
     * @param {boolean|undefined} includeInstance Deprecated. Whether to include
     *     the JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @param {!proto.TransactionProto} msg The msg instance to transform.
     * @return {!Object}
     * @suppress {unusedLocalVariables} f is only used for nested messages
     */
    proto.TransactionProto.toObject = function (includeInstance, msg) {
        var f,
            obj = {
                traceId:
                    (f = msg.getTraceId()) &&
                    proto.TraceID.toObject(includeInstance, f),
                logId: jspb.Message.getFieldWithDefault(msg, 2, ''),
                transactionId: jspb.Message.getFieldWithDefault(msg, 3, 0),
                version: jspb.Message.getFieldWithDefault(msg, 4, 0),
                rootSpanId: jspb.Message.getFieldWithDefault(msg, 5, '0'),
                flags: jspb.Message.getFieldWithDefault(msg, 6, 0),
                samplingType: jspb.Message.getFieldWithDefault(msg, 7, 0),
                samplingRate: jspb.Message.getFloatingPointFieldWithDefault(
                    msg,
                    8,
                    0.0
                ),
                debugInfo: jspb.Message.getFieldWithDefault(msg, 9, ''),
                customSpansList: jspb.Message.toObjectList(
                    msg.getCustomSpansList(),
                    proto.CustomSpanProto.toObject,
                    includeInstance
                ),
                serverSpansList: jspb.Message.toObjectList(
                    msg.getServerSpansList(),
                    proto.ServerSpanProto.toObject,
                    includeInstance
                ),
                clientSpansList: jspb.Message.toObjectList(
                    msg.getClientSpansList(),
                    proto.ClientSpanProto.toObject,
                    includeInstance
                ),
                service: jspb.Message.getFieldWithDefault(msg, 13, ''),
                dc: jspb.Message.getFieldWithDefault(msg, 14, ''),
                cluster: jspb.Message.getFieldWithDefault(msg, 15, ''),
                deployStage: jspb.Message.getFieldWithDefault(msg, 16, ''),
                env: jspb.Message.getFieldWithDefault(msg, 17, ''),
                podName: jspb.Message.getFieldWithDefault(msg, 18, ''),
                language: jspb.Message.getFieldWithDefault(msg, 19, ''),
                ipv4: jspb.Message.getFieldWithDefault(msg, 20, 0),
                ipv6: msg.getIpv6_asB64(),
                method: jspb.Message.getFieldWithDefault(msg, 22, ''),
                tagsList: jspb.Message.toObjectList(
                    msg.getTagsList(),
                    proto.Tag.toObject,
                    includeInstance
                ),
            };

        if (includeInstance) {
            obj.$jspbMessageInstance = msg;
        }
        return obj;
    };
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.TransactionProto}
 */
proto.TransactionProto.deserializeBinary = function (bytes) {
    var reader = new jspb.BinaryReader(bytes);
    var msg = new proto.TransactionProto();
    return proto.TransactionProto.deserializeBinaryFromReader(msg, reader);
};

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.TransactionProto} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.TransactionProto}
 */
proto.TransactionProto.deserializeBinaryFromReader = function (msg, reader) {
    while (reader.nextField()) {
        if (reader.isEndGroup()) {
            break;
        }
        var field = reader.getFieldNumber();
        switch (field) {
            case 1:
                var value = new proto.TraceID();
                reader.readMessage(
                    value,
                    proto.TraceID.deserializeBinaryFromReader
                );
                msg.setTraceId(value);
                break;
            case 2:
                var value = /** @type {string} */ (reader.readString());
                msg.setLogId(value);
                break;
            case 3:
                var value = /** @type {number} */ (reader.readUint64());
                msg.setTransactionId(value);
                break;
            case 4:
                var value = /** @type {number} */ (reader.readUint32());
                msg.setVersion(value);
                break;
            case 5:
                var value = /** @type {string} */ (reader.readUint64String());
                msg.setRootSpanId(value);
                break;
            case 6:
                var value = /** @type {number} */ (reader.readInt32());
                msg.setFlags(value);
                break;
            case 7:
                var value = /** @type {number} */ (reader.readInt32());
                msg.setSamplingType(value);
                break;
            case 8:
                var value = /** @type {number} */ (reader.readDouble());
                msg.setSamplingRate(value);
                break;
            case 9:
                var value = /** @type {string} */ (reader.readString());
                msg.setDebugInfo(value);
                break;
            case 10:
                var value = new proto.CustomSpanProto();
                reader.readMessage(
                    value,
                    proto.CustomSpanProto.deserializeBinaryFromReader
                );
                msg.addCustomSpans(value);
                break;
            case 11:
                var value = new proto.ServerSpanProto();
                reader.readMessage(
                    value,
                    proto.ServerSpanProto.deserializeBinaryFromReader
                );
                msg.addServerSpans(value);
                break;
            case 12:
                var value = new proto.ClientSpanProto();
                reader.readMessage(
                    value,
                    proto.ClientSpanProto.deserializeBinaryFromReader
                );
                msg.addClientSpans(value);
                break;
            case 13:
                var value = /** @type {string} */ (reader.readString());
                msg.setService(value);
                break;
            case 14:
                var value = /** @type {string} */ (reader.readString());
                msg.setDc(value);
                break;
            case 15:
                var value = /** @type {string} */ (reader.readString());
                msg.setCluster(value);
                break;
            case 16:
                var value = /** @type {string} */ (reader.readString());
                msg.setDeployStage(value);
                break;
            case 17:
                var value = /** @type {string} */ (reader.readString());
                msg.setEnv(value);
                break;
            case 18:
                var value = /** @type {string} */ (reader.readString());
                msg.setPodName(value);
                break;
            case 19:
                var value = /** @type {string} */ (reader.readString());
                msg.setLanguage(value);
                break;
            case 20:
                var value = /** @type {number} */ (reader.readUint32());
                msg.setIpv4(value);
                break;
            case 21:
                var value = /** @type {!Uint8Array} */ (reader.readBytes());
                msg.setIpv6(value);
                break;
            case 22:
                var value = /** @type {string} */ (reader.readString());
                msg.setMethod(value);
                break;
            case 23:
                var value = new proto.Tag();
                reader.readMessage(
                    value,
                    proto.Tag.deserializeBinaryFromReader
                );
                msg.addTags(value);
                break;
            default:
                reader.skipField();
                break;
        }
    }
    return msg;
};

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.TransactionProto.prototype.serializeBinary = function () {
    var writer = new jspb.BinaryWriter();
    proto.TransactionProto.serializeBinaryToWriter(this, writer);
    return writer.getResultBuffer();
};

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.TransactionProto} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.TransactionProto.serializeBinaryToWriter = function (message, writer) {
    var f = undefined;
    f = message.getTraceId();
    if (f != null) {
        writer.writeMessage(1, f, proto.TraceID.serializeBinaryToWriter);
    }
    f = message.getLogId();
    if (f.length > 0) {
        writer.writeString(2, f);
    }
    f = message.getTransactionId();
    if (f !== 0) {
        writer.writeUint64(3, f);
    }
    f = message.getVersion();
    if (f !== 0) {
        writer.writeUint32(4, f);
    }
    f = message.getRootSpanId();
    if (parseInt(f, 10) !== 0) {
        writer.writeUint64String(5, f);
    }
    f = message.getFlags();
    if (f !== 0) {
        writer.writeInt32(6, f);
    }
    f = message.getSamplingType();
    if (f !== 0) {
        writer.writeInt32(7, f);
    }
    f = message.getSamplingRate();
    if (f !== 0.0) {
        writer.writeDouble(8, f);
    }
    f = message.getDebugInfo();
    if (f.length > 0) {
        writer.writeString(9, f);
    }
    f = message.getCustomSpansList();
    if (f.length > 0) {
        writer.writeRepeatedMessage(
            10,
            f,
            proto.CustomSpanProto.serializeBinaryToWriter
        );
    }
    f = message.getServerSpansList();
    if (f.length > 0) {
        writer.writeRepeatedMessage(
            11,
            f,
            proto.ServerSpanProto.serializeBinaryToWriter
        );
    }
    f = message.getClientSpansList();
    if (f.length > 0) {
        writer.writeRepeatedMessage(
            12,
            f,
            proto.ClientSpanProto.serializeBinaryToWriter
        );
    }
    f = message.getService();
    if (f.length > 0) {
        writer.writeString(13, f);
    }
    f = message.getDc();
    if (f.length > 0) {
        writer.writeString(14, f);
    }
    f = message.getCluster();
    if (f.length > 0) {
        writer.writeString(15, f);
    }
    f = message.getDeployStage();
    if (f.length > 0) {
        writer.writeString(16, f);
    }
    f = message.getEnv();
    if (f.length > 0) {
        writer.writeString(17, f);
    }
    f = message.getPodName();
    if (f.length > 0) {
        writer.writeString(18, f);
    }
    f = message.getLanguage();
    if (f.length > 0) {
        writer.writeString(19, f);
    }
    f = message.getIpv4();
    if (f !== 0) {
        writer.writeUint32(20, f);
    }
    f = message.getIpv6_asU8();
    if (f.length > 0) {
        writer.writeBytes(21, f);
    }
    f = message.getMethod();
    if (f.length > 0) {
        writer.writeString(22, f);
    }
    f = message.getTagsList();
    if (f.length > 0) {
        writer.writeRepeatedMessage(23, f, proto.Tag.serializeBinaryToWriter);
    }
};

/**
 * optional TraceID trace_id = 1;
 * @return {?proto.TraceID}
 */
proto.TransactionProto.prototype.getTraceId = function () {
    return /** @type{?proto.TraceID} */ (
        jspb.Message.getWrapperField(this, proto.TraceID, 1)
    );
};

/**
 * @param {?proto.TraceID|undefined} value
 * @return {!proto.TransactionProto} returns this
 */
proto.TransactionProto.prototype.setTraceId = function (value) {
    return jspb.Message.setWrapperField(this, 1, value);
};

/**
 * Clears the message field making it undefined.
 * @return {!proto.TransactionProto} returns this
 */
proto.TransactionProto.prototype.clearTraceId = function () {
    return this.setTraceId(undefined);
};

/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.TransactionProto.prototype.hasTraceId = function () {
    return jspb.Message.getField(this, 1) != null;
};

/**
 * optional string log_id = 2;
 * @return {string}
 */
proto.TransactionProto.prototype.getLogId = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 2, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.TransactionProto} returns this
 */
proto.TransactionProto.prototype.setLogId = function (value) {
    return jspb.Message.setProto3StringField(this, 2, value);
};

/**
 * optional uint64 transaction_id = 3;
 * @return {number}
 */
proto.TransactionProto.prototype.getTransactionId = function () {
    return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 3, 0));
};

/**
 * @param {number} value
 * @return {!proto.TransactionProto} returns this
 */
proto.TransactionProto.prototype.setTransactionId = function (value) {
    return jspb.Message.setProto3IntField(this, 3, value);
};

/**
 * optional uint32 version = 4;
 * @return {number}
 */
proto.TransactionProto.prototype.getVersion = function () {
    return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 4, 0));
};

/**
 * @param {number} value
 * @return {!proto.TransactionProto} returns this
 */
proto.TransactionProto.prototype.setVersion = function (value) {
    return jspb.Message.setProto3IntField(this, 4, value);
};

/**
 * optional uint64 root_span_id = 5;
 * @return {string}
 */
proto.TransactionProto.prototype.getRootSpanId = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 5, '0')
    );
};

/**
 * @param {string} value
 * @return {!proto.TransactionProto} returns this
 */
proto.TransactionProto.prototype.setRootSpanId = function (value) {
    return jspb.Message.setProto3StringIntField(this, 5, value);
};

/**
 * optional int32 flags = 6;
 * @return {number}
 */
proto.TransactionProto.prototype.getFlags = function () {
    return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 6, 0));
};

/**
 * @param {number} value
 * @return {!proto.TransactionProto} returns this
 */
proto.TransactionProto.prototype.setFlags = function (value) {
    return jspb.Message.setProto3IntField(this, 6, value);
};

/**
 * optional int32 sampling_type = 7;
 * @return {number}
 */
proto.TransactionProto.prototype.getSamplingType = function () {
    return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 7, 0));
};

/**
 * @param {number} value
 * @return {!proto.TransactionProto} returns this
 */
proto.TransactionProto.prototype.setSamplingType = function (value) {
    return jspb.Message.setProto3IntField(this, 7, value);
};

/**
 * optional double sampling_rate = 8;
 * @return {number}
 */
proto.TransactionProto.prototype.getSamplingRate = function () {
    return /** @type {number} */ (
        jspb.Message.getFloatingPointFieldWithDefault(this, 8, 0.0)
    );
};

/**
 * @param {number} value
 * @return {!proto.TransactionProto} returns this
 */
proto.TransactionProto.prototype.setSamplingRate = function (value) {
    return jspb.Message.setProto3FloatField(this, 8, value);
};

/**
 * optional string debug_info = 9;
 * @return {string}
 */
proto.TransactionProto.prototype.getDebugInfo = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 9, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.TransactionProto} returns this
 */
proto.TransactionProto.prototype.setDebugInfo = function (value) {
    return jspb.Message.setProto3StringField(this, 9, value);
};

/**
 * repeated CustomSpanProto custom_spans = 10;
 * @return {!Array<!proto.CustomSpanProto>}
 */
proto.TransactionProto.prototype.getCustomSpansList = function () {
    return /** @type{!Array<!proto.CustomSpanProto>} */ (
        jspb.Message.getRepeatedWrapperField(this, proto.CustomSpanProto, 10)
    );
};

/**
 * @param {!Array<!proto.CustomSpanProto>} value
 * @return {!proto.TransactionProto} returns this
 */
proto.TransactionProto.prototype.setCustomSpansList = function (value) {
    return jspb.Message.setRepeatedWrapperField(this, 10, value);
};

/**
 * @param {!proto.CustomSpanProto=} opt_value
 * @param {number=} opt_index
 * @return {!proto.CustomSpanProto}
 */
proto.TransactionProto.prototype.addCustomSpans = function (
    opt_value,
    opt_index
) {
    return jspb.Message.addToRepeatedWrapperField(
        this,
        10,
        opt_value,
        proto.CustomSpanProto,
        opt_index
    );
};

/**
 * Clears the list making it empty but non-null.
 * @return {!proto.TransactionProto} returns this
 */
proto.TransactionProto.prototype.clearCustomSpansList = function () {
    return this.setCustomSpansList([]);
};

/**
 * repeated ServerSpanProto server_spans = 11;
 * @return {!Array<!proto.ServerSpanProto>}
 */
proto.TransactionProto.prototype.getServerSpansList = function () {
    return /** @type{!Array<!proto.ServerSpanProto>} */ (
        jspb.Message.getRepeatedWrapperField(this, proto.ServerSpanProto, 11)
    );
};

/**
 * @param {!Array<!proto.ServerSpanProto>} value
 * @return {!proto.TransactionProto} returns this
 */
proto.TransactionProto.prototype.setServerSpansList = function (value) {
    return jspb.Message.setRepeatedWrapperField(this, 11, value);
};

/**
 * @param {!proto.ServerSpanProto=} opt_value
 * @param {number=} opt_index
 * @return {!proto.ServerSpanProto}
 */
proto.TransactionProto.prototype.addServerSpans = function (
    opt_value,
    opt_index
) {
    return jspb.Message.addToRepeatedWrapperField(
        this,
        11,
        opt_value,
        proto.ServerSpanProto,
        opt_index
    );
};

/**
 * Clears the list making it empty but non-null.
 * @return {!proto.TransactionProto} returns this
 */
proto.TransactionProto.prototype.clearServerSpansList = function () {
    return this.setServerSpansList([]);
};

/**
 * repeated ClientSpanProto client_spans = 12;
 * @return {!Array<!proto.ClientSpanProto>}
 */
proto.TransactionProto.prototype.getClientSpansList = function () {
    return /** @type{!Array<!proto.ClientSpanProto>} */ (
        jspb.Message.getRepeatedWrapperField(this, proto.ClientSpanProto, 12)
    );
};

/**
 * @param {!Array<!proto.ClientSpanProto>} value
 * @return {!proto.TransactionProto} returns this
 */
proto.TransactionProto.prototype.setClientSpansList = function (value) {
    return jspb.Message.setRepeatedWrapperField(this, 12, value);
};

/**
 * @param {!proto.ClientSpanProto=} opt_value
 * @param {number=} opt_index
 * @return {!proto.ClientSpanProto}
 */
proto.TransactionProto.prototype.addClientSpans = function (
    opt_value,
    opt_index
) {
    return jspb.Message.addToRepeatedWrapperField(
        this,
        12,
        opt_value,
        proto.ClientSpanProto,
        opt_index
    );
};

/**
 * Clears the list making it empty but non-null.
 * @return {!proto.TransactionProto} returns this
 */
proto.TransactionProto.prototype.clearClientSpansList = function () {
    return this.setClientSpansList([]);
};

/**
 * optional string service = 13;
 * @return {string}
 */
proto.TransactionProto.prototype.getService = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 13, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.TransactionProto} returns this
 */
proto.TransactionProto.prototype.setService = function (value) {
    return jspb.Message.setProto3StringField(this, 13, value);
};

/**
 * optional string dc = 14;
 * @return {string}
 */
proto.TransactionProto.prototype.getDc = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 14, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.TransactionProto} returns this
 */
proto.TransactionProto.prototype.setDc = function (value) {
    return jspb.Message.setProto3StringField(this, 14, value);
};

/**
 * optional string cluster = 15;
 * @return {string}
 */
proto.TransactionProto.prototype.getCluster = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 15, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.TransactionProto} returns this
 */
proto.TransactionProto.prototype.setCluster = function (value) {
    return jspb.Message.setProto3StringField(this, 15, value);
};

/**
 * optional string deploy_stage = 16;
 * @return {string}
 */
proto.TransactionProto.prototype.getDeployStage = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 16, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.TransactionProto} returns this
 */
proto.TransactionProto.prototype.setDeployStage = function (value) {
    return jspb.Message.setProto3StringField(this, 16, value);
};

/**
 * optional string env = 17;
 * @return {string}
 */
proto.TransactionProto.prototype.getEnv = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 17, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.TransactionProto} returns this
 */
proto.TransactionProto.prototype.setEnv = function (value) {
    return jspb.Message.setProto3StringField(this, 17, value);
};

/**
 * optional string pod_name = 18;
 * @return {string}
 */
proto.TransactionProto.prototype.getPodName = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 18, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.TransactionProto} returns this
 */
proto.TransactionProto.prototype.setPodName = function (value) {
    return jspb.Message.setProto3StringField(this, 18, value);
};

/**
 * optional string language = 19;
 * @return {string}
 */
proto.TransactionProto.prototype.getLanguage = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 19, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.TransactionProto} returns this
 */
proto.TransactionProto.prototype.setLanguage = function (value) {
    return jspb.Message.setProto3StringField(this, 19, value);
};

/**
 * optional uint32 ipv4 = 20;
 * @return {number}
 */
proto.TransactionProto.prototype.getIpv4 = function () {
    return /** @type {number} */ (
        jspb.Message.getFieldWithDefault(this, 20, 0)
    );
};

/**
 * @param {number} value
 * @return {!proto.TransactionProto} returns this
 */
proto.TransactionProto.prototype.setIpv4 = function (value) {
    return jspb.Message.setProto3IntField(this, 20, value);
};

/**
 * optional bytes ipv6 = 21;
 * @return {!(string|Uint8Array)}
 */
proto.TransactionProto.prototype.getIpv6 = function () {
    return /** @type {!(string|Uint8Array)} */ (
        jspb.Message.getFieldWithDefault(this, 21, '')
    );
};

/**
 * optional bytes ipv6 = 21;
 * This is a type-conversion wrapper around `getIpv6()`
 * @return {string}
 */
proto.TransactionProto.prototype.getIpv6_asB64 = function () {
    return /** @type {string} */ (jspb.Message.bytesAsB64(this.getIpv6()));
};

/**
 * optional bytes ipv6 = 21;
 * Note that Uint8Array is not supported on all browsers.
 * @see http://caniuse.com/Uint8Array
 * This is a type-conversion wrapper around `getIpv6()`
 * @return {!Uint8Array}
 */
proto.TransactionProto.prototype.getIpv6_asU8 = function () {
    return /** @type {!Uint8Array} */ (jspb.Message.bytesAsU8(this.getIpv6()));
};

/**
 * @param {!(string|Uint8Array)} value
 * @return {!proto.TransactionProto} returns this
 */
proto.TransactionProto.prototype.setIpv6 = function (value) {
    return jspb.Message.setProto3BytesField(this, 21, value);
};

/**
 * optional string method = 22;
 * @return {string}
 */
proto.TransactionProto.prototype.getMethod = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 22, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.TransactionProto} returns this
 */
proto.TransactionProto.prototype.setMethod = function (value) {
    return jspb.Message.setProto3StringField(this, 22, value);
};

/**
 * repeated Tag tags = 23;
 * @return {!Array<!proto.Tag>}
 */
proto.TransactionProto.prototype.getTagsList = function () {
    return /** @type{!Array<!proto.Tag>} */ (
        jspb.Message.getRepeatedWrapperField(this, proto.Tag, 23)
    );
};

/**
 * @param {!Array<!proto.Tag>} value
 * @return {!proto.TransactionProto} returns this
 */
proto.TransactionProto.prototype.setTagsList = function (value) {
    return jspb.Message.setRepeatedWrapperField(this, 23, value);
};

/**
 * @param {!proto.Tag=} opt_value
 * @param {number=} opt_index
 * @return {!proto.Tag}
 */
proto.TransactionProto.prototype.addTags = function (opt_value, opt_index) {
    return jspb.Message.addToRepeatedWrapperField(
        this,
        23,
        opt_value,
        proto.Tag,
        opt_index
    );
};

/**
 * Clears the list making it empty but non-null.
 * @return {!proto.TransactionProto} returns this
 */
proto.TransactionProto.prototype.clearTagsList = function () {
    return this.setTagsList([]);
};

/**
 * List of repeated fields within this message type.
 * @private {!Array<number>}
 * @const
 */
proto.CustomSpanProto.repeatedFields_ = [4, 7, 8, 9];

if (jspb.Message.GENERATE_TO_OBJECT) {
    /**
     * Creates an object representation of this proto.
     * Field names that are reserved in JavaScript and will be renamed to pb_name.
     * Optional fields that are not set will be set to undefined.
     * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
     * For the list of reserved names please see:
     *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
     * @param {boolean=} opt_includeInstance Deprecated. whether to include the
     *     JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @return {!Object}
     */
    proto.CustomSpanProto.prototype.toObject = function (opt_includeInstance) {
        return proto.CustomSpanProto.toObject(opt_includeInstance, this);
    };

    /**
     * Static version of the {@see toObject} method.
     * @param {boolean|undefined} includeInstance Deprecated. Whether to include
     *     the JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @param {!proto.CustomSpanProto} msg The msg instance to transform.
     * @return {!Object}
     * @suppress {unusedLocalVariables} f is only used for nested messages
     */
    proto.CustomSpanProto.toObject = function (includeInstance, msg) {
        var f,
            obj = {
                spanId: jspb.Message.getFieldWithDefault(msg, 1, '0'),
                parentId: jspb.Message.getFieldWithDefault(msg, 2, '0'),
                refType: jspb.Message.getFieldWithDefault(msg, 3, 0),
                referencesList: jspb.Message.toObjectList(
                    msg.getReferencesList(),
                    proto.SpanRef.toObject,
                    includeInstance
                ),
                type: jspb.Message.getFieldWithDefault(msg, 5, ''),
                name: jspb.Message.getFieldWithDefault(msg, 6, ''),
                tagsList: jspb.Message.toObjectList(
                    msg.getTagsList(),
                    proto.Tag.toObject,
                    includeInstance
                ),
                eventsList: jspb.Message.toObjectList(
                    msg.getEventsList(),
                    proto.EventProto.toObject,
                    includeInstance
                ),
                metricsList: jspb.Message.toObjectList(
                    msg.getMetricsList(),
                    proto.MetricProto.toObject,
                    includeInstance
                ),
                duration: jspb.Message.getFieldWithDefault(msg, 10, 0),
                startTime: jspb.Message.getFieldWithDefault(msg, 11, 0),
                statusCode: jspb.Message.getFieldWithDefault(msg, 12, 0),
            };

        if (includeInstance) {
            obj.$jspbMessageInstance = msg;
        }
        return obj;
    };
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.CustomSpanProto}
 */
proto.CustomSpanProto.deserializeBinary = function (bytes) {
    var reader = new jspb.BinaryReader(bytes);
    var msg = new proto.CustomSpanProto();
    return proto.CustomSpanProto.deserializeBinaryFromReader(msg, reader);
};

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.CustomSpanProto} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.CustomSpanProto}
 */
proto.CustomSpanProto.deserializeBinaryFromReader = function (msg, reader) {
    while (reader.nextField()) {
        if (reader.isEndGroup()) {
            break;
        }
        var field = reader.getFieldNumber();
        switch (field) {
            case 1:
                var value = /** @type {string} */ (reader.readUint64String());
                msg.setSpanId(value);
                break;
            case 2:
                var value = /** @type {string} */ (reader.readUint64String());
                msg.setParentId(value);
                break;
            case 3:
                var value = /** @type {number} */ (reader.readInt32());
                msg.setRefType(value);
                break;
            case 4:
                var value = new proto.SpanRef();
                reader.readMessage(
                    value,
                    proto.SpanRef.deserializeBinaryFromReader
                );
                msg.addReferences(value);
                break;
            case 5:
                var value = /** @type {string} */ (reader.readString());
                msg.setType(value);
                break;
            case 6:
                var value = /** @type {string} */ (reader.readString());
                msg.setName(value);
                break;
            case 7:
                var value = new proto.Tag();
                reader.readMessage(
                    value,
                    proto.Tag.deserializeBinaryFromReader
                );
                msg.addTags(value);
                break;
            case 8:
                var value = new proto.EventProto();
                reader.readMessage(
                    value,
                    proto.EventProto.deserializeBinaryFromReader
                );
                msg.addEvents(value);
                break;
            case 9:
                var value = new proto.MetricProto();
                reader.readMessage(
                    value,
                    proto.MetricProto.deserializeBinaryFromReader
                );
                msg.addMetrics(value);
                break;
            case 10:
                var value = /** @type {number} */ (reader.readInt64());
                msg.setDuration(value);
                break;
            case 11:
                var value = /** @type {number} */ (reader.readInt64());
                msg.setStartTime(value);
                break;
            case 12:
                var value = /** @type {number} */ (reader.readInt32());
                msg.setStatusCode(value);
                break;
            default:
                reader.skipField();
                break;
        }
    }
    return msg;
};

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.CustomSpanProto.prototype.serializeBinary = function () {
    var writer = new jspb.BinaryWriter();
    proto.CustomSpanProto.serializeBinaryToWriter(this, writer);
    return writer.getResultBuffer();
};

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.CustomSpanProto} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.CustomSpanProto.serializeBinaryToWriter = function (message, writer) {
    var f = undefined;
    f = message.getSpanId();
    if (parseInt(f, 10) !== 0) {
        writer.writeUint64String(1, f);
    }
    f = message.getParentId();
    if (parseInt(f, 10) !== 0) {
        writer.writeUint64String(2, f);
    }
    f = message.getRefType();
    if (f !== 0) {
        writer.writeInt32(3, f);
    }
    f = message.getReferencesList();
    if (f.length > 0) {
        writer.writeRepeatedMessage(
            4,
            f,
            proto.SpanRef.serializeBinaryToWriter
        );
    }
    f = message.getType();
    if (f.length > 0) {
        writer.writeString(5, f);
    }
    f = message.getName();
    if (f.length > 0) {
        writer.writeString(6, f);
    }
    f = message.getTagsList();
    if (f.length > 0) {
        writer.writeRepeatedMessage(7, f, proto.Tag.serializeBinaryToWriter);
    }
    f = message.getEventsList();
    if (f.length > 0) {
        writer.writeRepeatedMessage(
            8,
            f,
            proto.EventProto.serializeBinaryToWriter
        );
    }
    f = message.getMetricsList();
    if (f.length > 0) {
        writer.writeRepeatedMessage(
            9,
            f,
            proto.MetricProto.serializeBinaryToWriter
        );
    }
    f = message.getDuration();
    if (f !== 0) {
        writer.writeInt64(10, f);
    }
    f = message.getStartTime();
    if (f !== 0) {
        writer.writeInt64(11, f);
    }
    f = message.getStatusCode();
    if (f !== 0) {
        writer.writeInt32(12, f);
    }
};

/**
 * optional uint64 span_id = 1;
 * @return {string}
 */
proto.CustomSpanProto.prototype.getSpanId = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 1, '0')
    );
};

/**
 * @param {string} value
 * @return {!proto.CustomSpanProto} returns this
 */
proto.CustomSpanProto.prototype.setSpanId = function (value) {
    return jspb.Message.setProto3StringIntField(this, 1, value);
};

/**
 * optional uint64 parent_id = 2;
 * @return {string}
 */
proto.CustomSpanProto.prototype.getParentId = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 2, '0')
    );
};

/**
 * @param {string} value
 * @return {!proto.CustomSpanProto} returns this
 */
proto.CustomSpanProto.prototype.setParentId = function (value) {
    return jspb.Message.setProto3StringIntField(this, 2, value);
};

/**
 * optional int32 ref_type = 3;
 * @return {number}
 */
proto.CustomSpanProto.prototype.getRefType = function () {
    return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 3, 0));
};

/**
 * @param {number} value
 * @return {!proto.CustomSpanProto} returns this
 */
proto.CustomSpanProto.prototype.setRefType = function (value) {
    return jspb.Message.setProto3IntField(this, 3, value);
};

/**
 * repeated SpanRef references = 4;
 * @return {!Array<!proto.SpanRef>}
 */
proto.CustomSpanProto.prototype.getReferencesList = function () {
    return /** @type{!Array<!proto.SpanRef>} */ (
        jspb.Message.getRepeatedWrapperField(this, proto.SpanRef, 4)
    );
};

/**
 * @param {!Array<!proto.SpanRef>} value
 * @return {!proto.CustomSpanProto} returns this
 */
proto.CustomSpanProto.prototype.setReferencesList = function (value) {
    return jspb.Message.setRepeatedWrapperField(this, 4, value);
};

/**
 * @param {!proto.SpanRef=} opt_value
 * @param {number=} opt_index
 * @return {!proto.SpanRef}
 */
proto.CustomSpanProto.prototype.addReferences = function (
    opt_value,
    opt_index
) {
    return jspb.Message.addToRepeatedWrapperField(
        this,
        4,
        opt_value,
        proto.SpanRef,
        opt_index
    );
};

/**
 * Clears the list making it empty but non-null.
 * @return {!proto.CustomSpanProto} returns this
 */
proto.CustomSpanProto.prototype.clearReferencesList = function () {
    return this.setReferencesList([]);
};

/**
 * optional string type = 5;
 * @return {string}
 */
proto.CustomSpanProto.prototype.getType = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 5, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.CustomSpanProto} returns this
 */
proto.CustomSpanProto.prototype.setType = function (value) {
    return jspb.Message.setProto3StringField(this, 5, value);
};

/**
 * optional string name = 6;
 * @return {string}
 */
proto.CustomSpanProto.prototype.getName = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 6, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.CustomSpanProto} returns this
 */
proto.CustomSpanProto.prototype.setName = function (value) {
    return jspb.Message.setProto3StringField(this, 6, value);
};

/**
 * repeated Tag tags = 7;
 * @return {!Array<!proto.Tag>}
 */
proto.CustomSpanProto.prototype.getTagsList = function () {
    return /** @type{!Array<!proto.Tag>} */ (
        jspb.Message.getRepeatedWrapperField(this, proto.Tag, 7)
    );
};

/**
 * @param {!Array<!proto.Tag>} value
 * @return {!proto.CustomSpanProto} returns this
 */
proto.CustomSpanProto.prototype.setTagsList = function (value) {
    return jspb.Message.setRepeatedWrapperField(this, 7, value);
};

/**
 * @param {!proto.Tag=} opt_value
 * @param {number=} opt_index
 * @return {!proto.Tag}
 */
proto.CustomSpanProto.prototype.addTags = function (opt_value, opt_index) {
    return jspb.Message.addToRepeatedWrapperField(
        this,
        7,
        opt_value,
        proto.Tag,
        opt_index
    );
};

/**
 * Clears the list making it empty but non-null.
 * @return {!proto.CustomSpanProto} returns this
 */
proto.CustomSpanProto.prototype.clearTagsList = function () {
    return this.setTagsList([]);
};

/**
 * repeated EventProto events = 8;
 * @return {!Array<!proto.EventProto>}
 */
proto.CustomSpanProto.prototype.getEventsList = function () {
    return /** @type{!Array<!proto.EventProto>} */ (
        jspb.Message.getRepeatedWrapperField(this, proto.EventProto, 8)
    );
};

/**
 * @param {!Array<!proto.EventProto>} value
 * @return {!proto.CustomSpanProto} returns this
 */
proto.CustomSpanProto.prototype.setEventsList = function (value) {
    return jspb.Message.setRepeatedWrapperField(this, 8, value);
};

/**
 * @param {!proto.EventProto=} opt_value
 * @param {number=} opt_index
 * @return {!proto.EventProto}
 */
proto.CustomSpanProto.prototype.addEvents = function (opt_value, opt_index) {
    return jspb.Message.addToRepeatedWrapperField(
        this,
        8,
        opt_value,
        proto.EventProto,
        opt_index
    );
};

/**
 * Clears the list making it empty but non-null.
 * @return {!proto.CustomSpanProto} returns this
 */
proto.CustomSpanProto.prototype.clearEventsList = function () {
    return this.setEventsList([]);
};

/**
 * repeated MetricProto metrics = 9;
 * @return {!Array<!proto.MetricProto>}
 */
proto.CustomSpanProto.prototype.getMetricsList = function () {
    return /** @type{!Array<!proto.MetricProto>} */ (
        jspb.Message.getRepeatedWrapperField(this, proto.MetricProto, 9)
    );
};

/**
 * @param {!Array<!proto.MetricProto>} value
 * @return {!proto.CustomSpanProto} returns this
 */
proto.CustomSpanProto.prototype.setMetricsList = function (value) {
    return jspb.Message.setRepeatedWrapperField(this, 9, value);
};

/**
 * @param {!proto.MetricProto=} opt_value
 * @param {number=} opt_index
 * @return {!proto.MetricProto}
 */
proto.CustomSpanProto.prototype.addMetrics = function (opt_value, opt_index) {
    return jspb.Message.addToRepeatedWrapperField(
        this,
        9,
        opt_value,
        proto.MetricProto,
        opt_index
    );
};

/**
 * Clears the list making it empty but non-null.
 * @return {!proto.CustomSpanProto} returns this
 */
proto.CustomSpanProto.prototype.clearMetricsList = function () {
    return this.setMetricsList([]);
};

/**
 * optional int64 duration = 10;
 * @return {number}
 */
proto.CustomSpanProto.prototype.getDuration = function () {
    return /** @type {number} */ (
        jspb.Message.getFieldWithDefault(this, 10, 0)
    );
};

/**
 * @param {number} value
 * @return {!proto.CustomSpanProto} returns this
 */
proto.CustomSpanProto.prototype.setDuration = function (value) {
    return jspb.Message.setProto3IntField(this, 10, value);
};

/**
 * optional int64 start_time = 11;
 * @return {number}
 */
proto.CustomSpanProto.prototype.getStartTime = function () {
    return /** @type {number} */ (
        jspb.Message.getFieldWithDefault(this, 11, 0)
    );
};

/**
 * @param {number} value
 * @return {!proto.CustomSpanProto} returns this
 */
proto.CustomSpanProto.prototype.setStartTime = function (value) {
    return jspb.Message.setProto3IntField(this, 11, value);
};

/**
 * optional int32 status_code = 12;
 * @return {number}
 */
proto.CustomSpanProto.prototype.getStatusCode = function () {
    return /** @type {number} */ (
        jspb.Message.getFieldWithDefault(this, 12, 0)
    );
};

/**
 * @param {number} value
 * @return {!proto.CustomSpanProto} returns this
 */
proto.CustomSpanProto.prototype.setStatusCode = function (value) {
    return jspb.Message.setProto3IntField(this, 12, value);
};

/**
 * List of repeated fields within this message type.
 * @private {!Array<number>}
 * @const
 */
proto.ServerSpanProto.repeatedFields_ = [4, 5, 6, 7];

if (jspb.Message.GENERATE_TO_OBJECT) {
    /**
     * Creates an object representation of this proto.
     * Field names that are reserved in JavaScript and will be renamed to pb_name.
     * Optional fields that are not set will be set to undefined.
     * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
     * For the list of reserved names please see:
     *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
     * @param {boolean=} opt_includeInstance Deprecated. whether to include the
     *     JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @return {!Object}
     */
    proto.ServerSpanProto.prototype.toObject = function (opt_includeInstance) {
        return proto.ServerSpanProto.toObject(opt_includeInstance, this);
    };

    /**
     * Static version of the {@see toObject} method.
     * @param {boolean|undefined} includeInstance Deprecated. Whether to include
     *     the JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @param {!proto.ServerSpanProto} msg The msg instance to transform.
     * @return {!Object}
     * @suppress {unusedLocalVariables} f is only used for nested messages
     */
    proto.ServerSpanProto.toObject = function (includeInstance, msg) {
        var f,
            obj = {
                spanId: jspb.Message.getFieldWithDefault(msg, 1, '0'),
                parentId: jspb.Message.getFieldWithDefault(msg, 2, '0'),
                refType: jspb.Message.getFieldWithDefault(msg, 3, 0),
                referencesList: jspb.Message.toObjectList(
                    msg.getReferencesList(),
                    proto.SpanRef.toObject,
                    includeInstance
                ),
                tagsList: jspb.Message.toObjectList(
                    msg.getTagsList(),
                    proto.Tag.toObject,
                    includeInstance
                ),
                eventsList: jspb.Message.toObjectList(
                    msg.getEventsList(),
                    proto.EventProto.toObject,
                    includeInstance
                ),
                metricsList: jspb.Message.toObjectList(
                    msg.getMetricsList(),
                    proto.MetricProto.toObject,
                    includeInstance
                ),
                duration: jspb.Message.getFieldWithDefault(msg, 8, 0),
                startTime: jspb.Message.getFieldWithDefault(msg, 9, 0),
                statusCode: jspb.Message.getFieldWithDefault(msg, 10, 0),
                method: jspb.Message.getFieldWithDefault(msg, 11, ''),
                component: jspb.Message.getFieldWithDefault(msg, 12, ''),
                serviceType: jspb.Message.getFieldWithDefault(msg, 13, ''),
                isError: jspb.Message.getBooleanFieldWithDefault(
                    msg,
                    14,
                    false
                ),
                businessStatusCode: jspb.Message.getFieldWithDefault(
                    msg,
                    15,
                    0
                ),
                fromService: jspb.Message.getFieldWithDefault(msg, 16, ''),
                fromMethod: jspb.Message.getFieldWithDefault(msg, 17, ''),
                fromCluster: jspb.Message.getFieldWithDefault(msg, 18, ''),
                fromAddr: jspb.Message.getFieldWithDefault(msg, 19, ''),
                fromDc: jspb.Message.getFieldWithDefault(msg, 20, ''),
                sendSize: jspb.Message.getFieldWithDefault(msg, 21, 0),
                recvSize: jspb.Message.getFieldWithDefault(msg, 22, 0),
            };

        if (includeInstance) {
            obj.$jspbMessageInstance = msg;
        }
        return obj;
    };
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.ServerSpanProto}
 */
proto.ServerSpanProto.deserializeBinary = function (bytes) {
    var reader = new jspb.BinaryReader(bytes);
    var msg = new proto.ServerSpanProto();
    return proto.ServerSpanProto.deserializeBinaryFromReader(msg, reader);
};

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.ServerSpanProto} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.ServerSpanProto}
 */
proto.ServerSpanProto.deserializeBinaryFromReader = function (msg, reader) {
    while (reader.nextField()) {
        if (reader.isEndGroup()) {
            break;
        }
        var field = reader.getFieldNumber();
        switch (field) {
            case 1:
                var value = /** @type {string} */ (reader.readUint64String());
                msg.setSpanId(value);
                break;
            case 2:
                var value = /** @type {string} */ (reader.readUint64String());
                msg.setParentId(value);
                break;
            case 3:
                var value = /** @type {number} */ (reader.readInt32());
                msg.setRefType(value);
                break;
            case 4:
                var value = new proto.SpanRef();
                reader.readMessage(
                    value,
                    proto.SpanRef.deserializeBinaryFromReader
                );
                msg.addReferences(value);
                break;
            case 5:
                var value = new proto.Tag();
                reader.readMessage(
                    value,
                    proto.Tag.deserializeBinaryFromReader
                );
                msg.addTags(value);
                break;
            case 6:
                var value = new proto.EventProto();
                reader.readMessage(
                    value,
                    proto.EventProto.deserializeBinaryFromReader
                );
                msg.addEvents(value);
                break;
            case 7:
                var value = new proto.MetricProto();
                reader.readMessage(
                    value,
                    proto.MetricProto.deserializeBinaryFromReader
                );
                msg.addMetrics(value);
                break;
            case 8:
                var value = /** @type {number} */ (reader.readInt64());
                msg.setDuration(value);
                break;
            case 9:
                var value = /** @type {number} */ (reader.readInt64());
                msg.setStartTime(value);
                break;
            case 10:
                var value = /** @type {number} */ (reader.readInt32());
                msg.setStatusCode(value);
                break;
            case 11:
                var value = /** @type {string} */ (reader.readString());
                msg.setMethod(value);
                break;
            case 12:
                var value = /** @type {string} */ (reader.readString());
                msg.setComponent(value);
                break;
            case 13:
                var value = /** @type {string} */ (reader.readString());
                msg.setServiceType(value);
                break;
            case 14:
                var value = /** @type {boolean} */ (reader.readBool());
                msg.setIsError(value);
                break;
            case 15:
                var value = /** @type {number} */ (reader.readInt32());
                msg.setBusinessStatusCode(value);
                break;
            case 16:
                var value = /** @type {string} */ (reader.readString());
                msg.setFromService(value);
                break;
            case 17:
                var value = /** @type {string} */ (reader.readString());
                msg.setFromMethod(value);
                break;
            case 18:
                var value = /** @type {string} */ (reader.readString());
                msg.setFromCluster(value);
                break;
            case 19:
                var value = /** @type {string} */ (reader.readString());
                msg.setFromAddr(value);
                break;
            case 20:
                var value = /** @type {string} */ (reader.readString());
                msg.setFromDc(value);
                break;
            case 21:
                var value = /** @type {number} */ (reader.readInt64());
                msg.setSendSize(value);
                break;
            case 22:
                var value = /** @type {number} */ (reader.readInt64());
                msg.setRecvSize(value);
                break;
            default:
                reader.skipField();
                break;
        }
    }
    return msg;
};

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.ServerSpanProto.prototype.serializeBinary = function () {
    var writer = new jspb.BinaryWriter();
    proto.ServerSpanProto.serializeBinaryToWriter(this, writer);
    return writer.getResultBuffer();
};

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.ServerSpanProto} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.ServerSpanProto.serializeBinaryToWriter = function (message, writer) {
    var f = undefined;
    f = message.getSpanId();
    if (parseInt(f, 10) !== 0) {
        writer.writeUint64String(1, f);
    }
    f = message.getParentId();
    if (parseInt(f, 10) !== 0) {
        writer.writeUint64String(2, f);
    }
    f = message.getRefType();
    if (f !== 0) {
        writer.writeInt32(3, f);
    }
    f = message.getReferencesList();
    if (f.length > 0) {
        writer.writeRepeatedMessage(
            4,
            f,
            proto.SpanRef.serializeBinaryToWriter
        );
    }
    f = message.getTagsList();
    if (f.length > 0) {
        writer.writeRepeatedMessage(5, f, proto.Tag.serializeBinaryToWriter);
    }
    f = message.getEventsList();
    if (f.length > 0) {
        writer.writeRepeatedMessage(
            6,
            f,
            proto.EventProto.serializeBinaryToWriter
        );
    }
    f = message.getMetricsList();
    if (f.length > 0) {
        writer.writeRepeatedMessage(
            7,
            f,
            proto.MetricProto.serializeBinaryToWriter
        );
    }
    f = message.getDuration();
    if (f !== 0) {
        writer.writeInt64(8, f);
    }
    f = message.getStartTime();
    if (f !== 0) {
        writer.writeInt64(9, f);
    }
    f = message.getStatusCode();
    if (f !== 0) {
        writer.writeInt32(10, f);
    }
    f = message.getMethod();
    if (f.length > 0) {
        writer.writeString(11, f);
    }
    f = message.getComponent();
    if (f.length > 0) {
        writer.writeString(12, f);
    }
    f = message.getServiceType();
    if (f.length > 0) {
        writer.writeString(13, f);
    }
    f = message.getIsError();
    if (f) {
        writer.writeBool(14, f);
    }
    f = message.getBusinessStatusCode();
    if (f !== 0) {
        writer.writeInt32(15, f);
    }
    f = message.getFromService();
    if (f.length > 0) {
        writer.writeString(16, f);
    }
    f = message.getFromMethod();
    if (f.length > 0) {
        writer.writeString(17, f);
    }
    f = message.getFromCluster();
    if (f.length > 0) {
        writer.writeString(18, f);
    }
    f = message.getFromAddr();
    if (f.length > 0) {
        writer.writeString(19, f);
    }
    f = message.getFromDc();
    if (f.length > 0) {
        writer.writeString(20, f);
    }
    f = message.getSendSize();
    if (f !== 0) {
        writer.writeInt64(21, f);
    }
    f = message.getRecvSize();
    if (f !== 0) {
        writer.writeInt64(22, f);
    }
};

/**
 * optional uint64 span_id = 1;
 * @return {string}
 */
proto.ServerSpanProto.prototype.getSpanId = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 1, '0')
    );
};

/**
 * @param {string} value
 * @return {!proto.ServerSpanProto} returns this
 */
proto.ServerSpanProto.prototype.setSpanId = function (value) {
    return jspb.Message.setProto3StringIntField(this, 1, value);
};

/**
 * optional uint64 parent_id = 2;
 * @return {string}
 */
proto.ServerSpanProto.prototype.getParentId = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 2, '0')
    );
};

/**
 * @param {string} value
 * @return {!proto.ServerSpanProto} returns this
 */
proto.ServerSpanProto.prototype.setParentId = function (value) {
    return jspb.Message.setProto3StringIntField(this, 2, value);
};

/**
 * optional int32 ref_type = 3;
 * @return {number}
 */
proto.ServerSpanProto.prototype.getRefType = function () {
    return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 3, 0));
};

/**
 * @param {number} value
 * @return {!proto.ServerSpanProto} returns this
 */
proto.ServerSpanProto.prototype.setRefType = function (value) {
    return jspb.Message.setProto3IntField(this, 3, value);
};

/**
 * repeated SpanRef references = 4;
 * @return {!Array<!proto.SpanRef>}
 */
proto.ServerSpanProto.prototype.getReferencesList = function () {
    return /** @type{!Array<!proto.SpanRef>} */ (
        jspb.Message.getRepeatedWrapperField(this, proto.SpanRef, 4)
    );
};

/**
 * @param {!Array<!proto.SpanRef>} value
 * @return {!proto.ServerSpanProto} returns this
 */
proto.ServerSpanProto.prototype.setReferencesList = function (value) {
    return jspb.Message.setRepeatedWrapperField(this, 4, value);
};

/**
 * @param {!proto.SpanRef=} opt_value
 * @param {number=} opt_index
 * @return {!proto.SpanRef}
 */
proto.ServerSpanProto.prototype.addReferences = function (
    opt_value,
    opt_index
) {
    return jspb.Message.addToRepeatedWrapperField(
        this,
        4,
        opt_value,
        proto.SpanRef,
        opt_index
    );
};

/**
 * Clears the list making it empty but non-null.
 * @return {!proto.ServerSpanProto} returns this
 */
proto.ServerSpanProto.prototype.clearReferencesList = function () {
    return this.setReferencesList([]);
};

/**
 * repeated Tag tags = 5;
 * @return {!Array<!proto.Tag>}
 */
proto.ServerSpanProto.prototype.getTagsList = function () {
    return /** @type{!Array<!proto.Tag>} */ (
        jspb.Message.getRepeatedWrapperField(this, proto.Tag, 5)
    );
};

/**
 * @param {!Array<!proto.Tag>} value
 * @return {!proto.ServerSpanProto} returns this
 */
proto.ServerSpanProto.prototype.setTagsList = function (value) {
    return jspb.Message.setRepeatedWrapperField(this, 5, value);
};

/**
 * @param {!proto.Tag=} opt_value
 * @param {number=} opt_index
 * @return {!proto.Tag}
 */
proto.ServerSpanProto.prototype.addTags = function (opt_value, opt_index) {
    return jspb.Message.addToRepeatedWrapperField(
        this,
        5,
        opt_value,
        proto.Tag,
        opt_index
    );
};

/**
 * Clears the list making it empty but non-null.
 * @return {!proto.ServerSpanProto} returns this
 */
proto.ServerSpanProto.prototype.clearTagsList = function () {
    return this.setTagsList([]);
};

/**
 * repeated EventProto events = 6;
 * @return {!Array<!proto.EventProto>}
 */
proto.ServerSpanProto.prototype.getEventsList = function () {
    return /** @type{!Array<!proto.EventProto>} */ (
        jspb.Message.getRepeatedWrapperField(this, proto.EventProto, 6)
    );
};

/**
 * @param {!Array<!proto.EventProto>} value
 * @return {!proto.ServerSpanProto} returns this
 */
proto.ServerSpanProto.prototype.setEventsList = function (value) {
    return jspb.Message.setRepeatedWrapperField(this, 6, value);
};

/**
 * @param {!proto.EventProto=} opt_value
 * @param {number=} opt_index
 * @return {!proto.EventProto}
 */
proto.ServerSpanProto.prototype.addEvents = function (opt_value, opt_index) {
    return jspb.Message.addToRepeatedWrapperField(
        this,
        6,
        opt_value,
        proto.EventProto,
        opt_index
    );
};

/**
 * Clears the list making it empty but non-null.
 * @return {!proto.ServerSpanProto} returns this
 */
proto.ServerSpanProto.prototype.clearEventsList = function () {
    return this.setEventsList([]);
};

/**
 * repeated MetricProto metrics = 7;
 * @return {!Array<!proto.MetricProto>}
 */
proto.ServerSpanProto.prototype.getMetricsList = function () {
    return /** @type{!Array<!proto.MetricProto>} */ (
        jspb.Message.getRepeatedWrapperField(this, proto.MetricProto, 7)
    );
};

/**
 * @param {!Array<!proto.MetricProto>} value
 * @return {!proto.ServerSpanProto} returns this
 */
proto.ServerSpanProto.prototype.setMetricsList = function (value) {
    return jspb.Message.setRepeatedWrapperField(this, 7, value);
};

/**
 * @param {!proto.MetricProto=} opt_value
 * @param {number=} opt_index
 * @return {!proto.MetricProto}
 */
proto.ServerSpanProto.prototype.addMetrics = function (opt_value, opt_index) {
    return jspb.Message.addToRepeatedWrapperField(
        this,
        7,
        opt_value,
        proto.MetricProto,
        opt_index
    );
};

/**
 * Clears the list making it empty but non-null.
 * @return {!proto.ServerSpanProto} returns this
 */
proto.ServerSpanProto.prototype.clearMetricsList = function () {
    return this.setMetricsList([]);
};

/**
 * optional int64 duration = 8;
 * @return {number}
 */
proto.ServerSpanProto.prototype.getDuration = function () {
    return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 8, 0));
};

/**
 * @param {number} value
 * @return {!proto.ServerSpanProto} returns this
 */
proto.ServerSpanProto.prototype.setDuration = function (value) {
    return jspb.Message.setProto3IntField(this, 8, value);
};

/**
 * optional int64 start_time = 9;
 * @return {number}
 */
proto.ServerSpanProto.prototype.getStartTime = function () {
    return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 9, 0));
};

/**
 * @param {number} value
 * @return {!proto.ServerSpanProto} returns this
 */
proto.ServerSpanProto.prototype.setStartTime = function (value) {
    return jspb.Message.setProto3IntField(this, 9, value);
};

/**
 * optional int32 status_code = 10;
 * @return {number}
 */
proto.ServerSpanProto.prototype.getStatusCode = function () {
    return /** @type {number} */ (
        jspb.Message.getFieldWithDefault(this, 10, 0)
    );
};

/**
 * @param {number} value
 * @return {!proto.ServerSpanProto} returns this
 */
proto.ServerSpanProto.prototype.setStatusCode = function (value) {
    return jspb.Message.setProto3IntField(this, 10, value);
};

/**
 * optional string method = 11;
 * @return {string}
 */
proto.ServerSpanProto.prototype.getMethod = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 11, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.ServerSpanProto} returns this
 */
proto.ServerSpanProto.prototype.setMethod = function (value) {
    return jspb.Message.setProto3StringField(this, 11, value);
};

/**
 * optional string component = 12;
 * @return {string}
 */
proto.ServerSpanProto.prototype.getComponent = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 12, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.ServerSpanProto} returns this
 */
proto.ServerSpanProto.prototype.setComponent = function (value) {
    return jspb.Message.setProto3StringField(this, 12, value);
};

/**
 * optional string service_type = 13;
 * @return {string}
 */
proto.ServerSpanProto.prototype.getServiceType = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 13, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.ServerSpanProto} returns this
 */
proto.ServerSpanProto.prototype.setServiceType = function (value) {
    return jspb.Message.setProto3StringField(this, 13, value);
};

/**
 * optional bool is_error = 14;
 * @return {boolean}
 */
proto.ServerSpanProto.prototype.getIsError = function () {
    return /** @type {boolean} */ (
        jspb.Message.getBooleanFieldWithDefault(this, 14, false)
    );
};

/**
 * @param {boolean} value
 * @return {!proto.ServerSpanProto} returns this
 */
proto.ServerSpanProto.prototype.setIsError = function (value) {
    return jspb.Message.setProto3BooleanField(this, 14, value);
};

/**
 * optional int32 business_status_code = 15;
 * @return {number}
 */
proto.ServerSpanProto.prototype.getBusinessStatusCode = function () {
    return /** @type {number} */ (
        jspb.Message.getFieldWithDefault(this, 15, 0)
    );
};

/**
 * @param {number} value
 * @return {!proto.ServerSpanProto} returns this
 */
proto.ServerSpanProto.prototype.setBusinessStatusCode = function (value) {
    return jspb.Message.setProto3IntField(this, 15, value);
};

/**
 * optional string from_service = 16;
 * @return {string}
 */
proto.ServerSpanProto.prototype.getFromService = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 16, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.ServerSpanProto} returns this
 */
proto.ServerSpanProto.prototype.setFromService = function (value) {
    return jspb.Message.setProto3StringField(this, 16, value);
};

/**
 * optional string from_method = 17;
 * @return {string}
 */
proto.ServerSpanProto.prototype.getFromMethod = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 17, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.ServerSpanProto} returns this
 */
proto.ServerSpanProto.prototype.setFromMethod = function (value) {
    return jspb.Message.setProto3StringField(this, 17, value);
};

/**
 * optional string from_cluster = 18;
 * @return {string}
 */
proto.ServerSpanProto.prototype.getFromCluster = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 18, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.ServerSpanProto} returns this
 */
proto.ServerSpanProto.prototype.setFromCluster = function (value) {
    return jspb.Message.setProto3StringField(this, 18, value);
};

/**
 * optional string from_addr = 19;
 * @return {string}
 */
proto.ServerSpanProto.prototype.getFromAddr = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 19, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.ServerSpanProto} returns this
 */
proto.ServerSpanProto.prototype.setFromAddr = function (value) {
    return jspb.Message.setProto3StringField(this, 19, value);
};

/**
 * optional string from_dc = 20;
 * @return {string}
 */
proto.ServerSpanProto.prototype.getFromDc = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 20, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.ServerSpanProto} returns this
 */
proto.ServerSpanProto.prototype.setFromDc = function (value) {
    return jspb.Message.setProto3StringField(this, 20, value);
};

/**
 * optional int64 send_size = 21;
 * @return {number}
 */
proto.ServerSpanProto.prototype.getSendSize = function () {
    return /** @type {number} */ (
        jspb.Message.getFieldWithDefault(this, 21, 0)
    );
};

/**
 * @param {number} value
 * @return {!proto.ServerSpanProto} returns this
 */
proto.ServerSpanProto.prototype.setSendSize = function (value) {
    return jspb.Message.setProto3IntField(this, 21, value);
};

/**
 * optional int64 recv_size = 22;
 * @return {number}
 */
proto.ServerSpanProto.prototype.getRecvSize = function () {
    return /** @type {number} */ (
        jspb.Message.getFieldWithDefault(this, 22, 0)
    );
};

/**
 * @param {number} value
 * @return {!proto.ServerSpanProto} returns this
 */
proto.ServerSpanProto.prototype.setRecvSize = function (value) {
    return jspb.Message.setProto3IntField(this, 22, value);
};

/**
 * List of repeated fields within this message type.
 * @private {!Array<number>}
 * @const
 */
proto.ClientSpanProto.repeatedFields_ = [4, 5, 6, 7];

if (jspb.Message.GENERATE_TO_OBJECT) {
    /**
     * Creates an object representation of this proto.
     * Field names that are reserved in JavaScript and will be renamed to pb_name.
     * Optional fields that are not set will be set to undefined.
     * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
     * For the list of reserved names please see:
     *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
     * @param {boolean=} opt_includeInstance Deprecated. whether to include the
     *     JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @return {!Object}
     */
    proto.ClientSpanProto.prototype.toObject = function (opt_includeInstance) {
        return proto.ClientSpanProto.toObject(opt_includeInstance, this);
    };

    /**
     * Static version of the {@see toObject} method.
     * @param {boolean|undefined} includeInstance Deprecated. Whether to include
     *     the JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @param {!proto.ClientSpanProto} msg The msg instance to transform.
     * @return {!Object}
     * @suppress {unusedLocalVariables} f is only used for nested messages
     */
    proto.ClientSpanProto.toObject = function (includeInstance, msg) {
        var f,
            obj = {
                spanId: jspb.Message.getFieldWithDefault(msg, 1, '0'),
                parentId: jspb.Message.getFieldWithDefault(msg, 2, '0'),
                refType: jspb.Message.getFieldWithDefault(msg, 3, 0),
                referencesList: jspb.Message.toObjectList(
                    msg.getReferencesList(),
                    proto.SpanRef.toObject,
                    includeInstance
                ),
                tagsList: jspb.Message.toObjectList(
                    msg.getTagsList(),
                    proto.Tag.toObject,
                    includeInstance
                ),
                eventsList: jspb.Message.toObjectList(
                    msg.getEventsList(),
                    proto.EventProto.toObject,
                    includeInstance
                ),
                metricsList: jspb.Message.toObjectList(
                    msg.getMetricsList(),
                    proto.MetricProto.toObject,
                    includeInstance
                ),
                duration: jspb.Message.getFieldWithDefault(msg, 8, 0),
                startTime: jspb.Message.getFieldWithDefault(msg, 9, 0),
                statusCode: jspb.Message.getFieldWithDefault(msg, 10, 0),
                toService: jspb.Message.getFieldWithDefault(msg, 11, ''),
                component: jspb.Message.getFieldWithDefault(msg, 12, ''),
                toServiceType: jspb.Message.getFieldWithDefault(msg, 13, ''),
                isError: jspb.Message.getBooleanFieldWithDefault(
                    msg,
                    14,
                    false
                ),
                businessStatusCode: jspb.Message.getFieldWithDefault(
                    msg,
                    15,
                    0
                ),
                toMethod: jspb.Message.getFieldWithDefault(msg, 16, ''),
                toCluster: jspb.Message.getFieldWithDefault(msg, 17, ''),
                toAddr: jspb.Message.getFieldWithDefault(msg, 18, ''),
                toDc: jspb.Message.getFieldWithDefault(msg, 19, ''),
                sendSize: jspb.Message.getFieldWithDefault(msg, 20, 0),
                recvSize: jspb.Message.getFieldWithDefault(msg, 21, 0),
            };

        if (includeInstance) {
            obj.$jspbMessageInstance = msg;
        }
        return obj;
    };
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.ClientSpanProto}
 */
proto.ClientSpanProto.deserializeBinary = function (bytes) {
    var reader = new jspb.BinaryReader(bytes);
    var msg = new proto.ClientSpanProto();
    return proto.ClientSpanProto.deserializeBinaryFromReader(msg, reader);
};

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.ClientSpanProto} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.ClientSpanProto}
 */
proto.ClientSpanProto.deserializeBinaryFromReader = function (msg, reader) {
    while (reader.nextField()) {
        if (reader.isEndGroup()) {
            break;
        }
        var field = reader.getFieldNumber();
        switch (field) {
            case 1:
                var value = /** @type {string} */ (reader.readUint64String());
                msg.setSpanId(value);
                break;
            case 2:
                var value = /** @type {string} */ (reader.readUint64String());
                msg.setParentId(value);
                break;
            case 3:
                var value = /** @type {number} */ (reader.readInt32());
                msg.setRefType(value);
                break;
            case 4:
                var value = new proto.SpanRef();
                reader.readMessage(
                    value,
                    proto.SpanRef.deserializeBinaryFromReader
                );
                msg.addReferences(value);
                break;
            case 5:
                var value = new proto.Tag();
                reader.readMessage(
                    value,
                    proto.Tag.deserializeBinaryFromReader
                );
                msg.addTags(value);
                break;
            case 6:
                var value = new proto.EventProto();
                reader.readMessage(
                    value,
                    proto.EventProto.deserializeBinaryFromReader
                );
                msg.addEvents(value);
                break;
            case 7:
                var value = new proto.MetricProto();
                reader.readMessage(
                    value,
                    proto.MetricProto.deserializeBinaryFromReader
                );
                msg.addMetrics(value);
                break;
            case 8:
                var value = /** @type {number} */ (reader.readInt64());
                msg.setDuration(value);
                break;
            case 9:
                var value = /** @type {number} */ (reader.readInt64());
                msg.setStartTime(value);
                break;
            case 10:
                var value = /** @type {number} */ (reader.readInt32());
                msg.setStatusCode(value);
                break;
            case 11:
                var value = /** @type {string} */ (reader.readString());
                msg.setToService(value);
                break;
            case 12:
                var value = /** @type {string} */ (reader.readString());
                msg.setComponent(value);
                break;
            case 13:
                var value = /** @type {string} */ (reader.readString());
                msg.setToServiceType(value);
                break;
            case 14:
                var value = /** @type {boolean} */ (reader.readBool());
                msg.setIsError(value);
                break;
            case 15:
                var value = /** @type {number} */ (reader.readInt32());
                msg.setBusinessStatusCode(value);
                break;
            case 16:
                var value = /** @type {string} */ (reader.readString());
                msg.setToMethod(value);
                break;
            case 17:
                var value = /** @type {string} */ (reader.readString());
                msg.setToCluster(value);
                break;
            case 18:
                var value = /** @type {string} */ (reader.readString());
                msg.setToAddr(value);
                break;
            case 19:
                var value = /** @type {string} */ (reader.readString());
                msg.setToDc(value);
                break;
            case 20:
                var value = /** @type {number} */ (reader.readInt64());
                msg.setSendSize(value);
                break;
            case 21:
                var value = /** @type {number} */ (reader.readInt64());
                msg.setRecvSize(value);
                break;
            default:
                reader.skipField();
                break;
        }
    }
    return msg;
};

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.ClientSpanProto.prototype.serializeBinary = function () {
    var writer = new jspb.BinaryWriter();
    proto.ClientSpanProto.serializeBinaryToWriter(this, writer);
    return writer.getResultBuffer();
};

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.ClientSpanProto} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.ClientSpanProto.serializeBinaryToWriter = function (message, writer) {
    var f = undefined;
    f = message.getSpanId();
    if (parseInt(f, 10) !== 0) {
        writer.writeUint64String(1, f);
    }
    f = message.getParentId();
    if (parseInt(f, 10) !== 0) {
        writer.writeUint64String(2, f);
    }
    f = message.getRefType();
    if (f !== 0) {
        writer.writeInt32(3, f);
    }
    f = message.getReferencesList();
    if (f.length > 0) {
        writer.writeRepeatedMessage(
            4,
            f,
            proto.SpanRef.serializeBinaryToWriter
        );
    }
    f = message.getTagsList();
    if (f.length > 0) {
        writer.writeRepeatedMessage(5, f, proto.Tag.serializeBinaryToWriter);
    }
    f = message.getEventsList();
    if (f.length > 0) {
        writer.writeRepeatedMessage(
            6,
            f,
            proto.EventProto.serializeBinaryToWriter
        );
    }
    f = message.getMetricsList();
    if (f.length > 0) {
        writer.writeRepeatedMessage(
            7,
            f,
            proto.MetricProto.serializeBinaryToWriter
        );
    }
    f = message.getDuration();
    if (f !== 0) {
        writer.writeInt64(8, f);
    }
    f = message.getStartTime();
    if (f !== 0) {
        writer.writeInt64(9, f);
    }
    f = message.getStatusCode();
    if (f !== 0) {
        writer.writeInt32(10, f);
    }
    f = message.getToService();
    if (f.length > 0) {
        writer.writeString(11, f);
    }
    f = message.getComponent();
    if (f.length > 0) {
        writer.writeString(12, f);
    }
    f = message.getToServiceType();
    if (f.length > 0) {
        writer.writeString(13, f);
    }
    f = message.getIsError();
    if (f) {
        writer.writeBool(14, f);
    }
    f = message.getBusinessStatusCode();
    if (f !== 0) {
        writer.writeInt32(15, f);
    }
    f = message.getToMethod();
    if (f.length > 0) {
        writer.writeString(16, f);
    }
    f = message.getToCluster();
    if (f.length > 0) {
        writer.writeString(17, f);
    }
    f = message.getToAddr();
    if (f.length > 0) {
        writer.writeString(18, f);
    }
    f = message.getToDc();
    if (f.length > 0) {
        writer.writeString(19, f);
    }
    f = message.getSendSize();
    if (f !== 0) {
        writer.writeInt64(20, f);
    }
    f = message.getRecvSize();
    if (f !== 0) {
        writer.writeInt64(21, f);
    }
};

/**
 * optional uint64 span_id = 1;
 * @return {string}
 */
proto.ClientSpanProto.prototype.getSpanId = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 1, '0')
    );
};

/**
 * @param {string} value
 * @return {!proto.ClientSpanProto} returns this
 */
proto.ClientSpanProto.prototype.setSpanId = function (value) {
    return jspb.Message.setProto3StringIntField(this, 1, value);
};

/**
 * optional uint64 parent_id = 2;
 * @return {string}
 */
proto.ClientSpanProto.prototype.getParentId = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 2, '0')
    );
};

/**
 * @param {string} value
 * @return {!proto.ClientSpanProto} returns this
 */
proto.ClientSpanProto.prototype.setParentId = function (value) {
    return jspb.Message.setProto3StringIntField(this, 2, value);
};

/**
 * optional int32 ref_type = 3;
 * @return {number}
 */
proto.ClientSpanProto.prototype.getRefType = function () {
    return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 3, 0));
};

/**
 * @param {number} value
 * @return {!proto.ClientSpanProto} returns this
 */
proto.ClientSpanProto.prototype.setRefType = function (value) {
    return jspb.Message.setProto3IntField(this, 3, value);
};

/**
 * repeated SpanRef references = 4;
 * @return {!Array<!proto.SpanRef>}
 */
proto.ClientSpanProto.prototype.getReferencesList = function () {
    return /** @type{!Array<!proto.SpanRef>} */ (
        jspb.Message.getRepeatedWrapperField(this, proto.SpanRef, 4)
    );
};

/**
 * @param {!Array<!proto.SpanRef>} value
 * @return {!proto.ClientSpanProto} returns this
 */
proto.ClientSpanProto.prototype.setReferencesList = function (value) {
    return jspb.Message.setRepeatedWrapperField(this, 4, value);
};

/**
 * @param {!proto.SpanRef=} opt_value
 * @param {number=} opt_index
 * @return {!proto.SpanRef}
 */
proto.ClientSpanProto.prototype.addReferences = function (
    opt_value,
    opt_index
) {
    return jspb.Message.addToRepeatedWrapperField(
        this,
        4,
        opt_value,
        proto.SpanRef,
        opt_index
    );
};

/**
 * Clears the list making it empty but non-null.
 * @return {!proto.ClientSpanProto} returns this
 */
proto.ClientSpanProto.prototype.clearReferencesList = function () {
    return this.setReferencesList([]);
};

/**
 * repeated Tag tags = 5;
 * @return {!Array<!proto.Tag>}
 */
proto.ClientSpanProto.prototype.getTagsList = function () {
    return /** @type{!Array<!proto.Tag>} */ (
        jspb.Message.getRepeatedWrapperField(this, proto.Tag, 5)
    );
};

/**
 * @param {!Array<!proto.Tag>} value
 * @return {!proto.ClientSpanProto} returns this
 */
proto.ClientSpanProto.prototype.setTagsList = function (value) {
    return jspb.Message.setRepeatedWrapperField(this, 5, value);
};

/**
 * @param {!proto.Tag=} opt_value
 * @param {number=} opt_index
 * @return {!proto.Tag}
 */
proto.ClientSpanProto.prototype.addTags = function (opt_value, opt_index) {
    return jspb.Message.addToRepeatedWrapperField(
        this,
        5,
        opt_value,
        proto.Tag,
        opt_index
    );
};

/**
 * Clears the list making it empty but non-null.
 * @return {!proto.ClientSpanProto} returns this
 */
proto.ClientSpanProto.prototype.clearTagsList = function () {
    return this.setTagsList([]);
};

/**
 * repeated EventProto events = 6;
 * @return {!Array<!proto.EventProto>}
 */
proto.ClientSpanProto.prototype.getEventsList = function () {
    return /** @type{!Array<!proto.EventProto>} */ (
        jspb.Message.getRepeatedWrapperField(this, proto.EventProto, 6)
    );
};

/**
 * @param {!Array<!proto.EventProto>} value
 * @return {!proto.ClientSpanProto} returns this
 */
proto.ClientSpanProto.prototype.setEventsList = function (value) {
    return jspb.Message.setRepeatedWrapperField(this, 6, value);
};

/**
 * @param {!proto.EventProto=} opt_value
 * @param {number=} opt_index
 * @return {!proto.EventProto}
 */
proto.ClientSpanProto.prototype.addEvents = function (opt_value, opt_index) {
    return jspb.Message.addToRepeatedWrapperField(
        this,
        6,
        opt_value,
        proto.EventProto,
        opt_index
    );
};

/**
 * Clears the list making it empty but non-null.
 * @return {!proto.ClientSpanProto} returns this
 */
proto.ClientSpanProto.prototype.clearEventsList = function () {
    return this.setEventsList([]);
};

/**
 * repeated MetricProto metrics = 7;
 * @return {!Array<!proto.MetricProto>}
 */
proto.ClientSpanProto.prototype.getMetricsList = function () {
    return /** @type{!Array<!proto.MetricProto>} */ (
        jspb.Message.getRepeatedWrapperField(this, proto.MetricProto, 7)
    );
};

/**
 * @param {!Array<!proto.MetricProto>} value
 * @return {!proto.ClientSpanProto} returns this
 */
proto.ClientSpanProto.prototype.setMetricsList = function (value) {
    return jspb.Message.setRepeatedWrapperField(this, 7, value);
};

/**
 * @param {!proto.MetricProto=} opt_value
 * @param {number=} opt_index
 * @return {!proto.MetricProto}
 */
proto.ClientSpanProto.prototype.addMetrics = function (opt_value, opt_index) {
    return jspb.Message.addToRepeatedWrapperField(
        this,
        7,
        opt_value,
        proto.MetricProto,
        opt_index
    );
};

/**
 * Clears the list making it empty but non-null.
 * @return {!proto.ClientSpanProto} returns this
 */
proto.ClientSpanProto.prototype.clearMetricsList = function () {
    return this.setMetricsList([]);
};

/**
 * optional int64 duration = 8;
 * @return {number}
 */
proto.ClientSpanProto.prototype.getDuration = function () {
    return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 8, 0));
};

/**
 * @param {number} value
 * @return {!proto.ClientSpanProto} returns this
 */
proto.ClientSpanProto.prototype.setDuration = function (value) {
    return jspb.Message.setProto3IntField(this, 8, value);
};

/**
 * optional int64 start_time = 9;
 * @return {number}
 */
proto.ClientSpanProto.prototype.getStartTime = function () {
    return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 9, 0));
};

/**
 * @param {number} value
 * @return {!proto.ClientSpanProto} returns this
 */
proto.ClientSpanProto.prototype.setStartTime = function (value) {
    return jspb.Message.setProto3IntField(this, 9, value);
};

/**
 * optional int32 status_code = 10;
 * @return {number}
 */
proto.ClientSpanProto.prototype.getStatusCode = function () {
    return /** @type {number} */ (
        jspb.Message.getFieldWithDefault(this, 10, 0)
    );
};

/**
 * @param {number} value
 * @return {!proto.ClientSpanProto} returns this
 */
proto.ClientSpanProto.prototype.setStatusCode = function (value) {
    return jspb.Message.setProto3IntField(this, 10, value);
};

/**
 * optional string to_service = 11;
 * @return {string}
 */
proto.ClientSpanProto.prototype.getToService = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 11, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.ClientSpanProto} returns this
 */
proto.ClientSpanProto.prototype.setToService = function (value) {
    return jspb.Message.setProto3StringField(this, 11, value);
};

/**
 * optional string component = 12;
 * @return {string}
 */
proto.ClientSpanProto.prototype.getComponent = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 12, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.ClientSpanProto} returns this
 */
proto.ClientSpanProto.prototype.setComponent = function (value) {
    return jspb.Message.setProto3StringField(this, 12, value);
};

/**
 * optional string to_service_type = 13;
 * @return {string}
 */
proto.ClientSpanProto.prototype.getToServiceType = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 13, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.ClientSpanProto} returns this
 */
proto.ClientSpanProto.prototype.setToServiceType = function (value) {
    return jspb.Message.setProto3StringField(this, 13, value);
};

/**
 * optional bool is_error = 14;
 * @return {boolean}
 */
proto.ClientSpanProto.prototype.getIsError = function () {
    return /** @type {boolean} */ (
        jspb.Message.getBooleanFieldWithDefault(this, 14, false)
    );
};

/**
 * @param {boolean} value
 * @return {!proto.ClientSpanProto} returns this
 */
proto.ClientSpanProto.prototype.setIsError = function (value) {
    return jspb.Message.setProto3BooleanField(this, 14, value);
};

/**
 * optional int32 business_status_code = 15;
 * @return {number}
 */
proto.ClientSpanProto.prototype.getBusinessStatusCode = function () {
    return /** @type {number} */ (
        jspb.Message.getFieldWithDefault(this, 15, 0)
    );
};

/**
 * @param {number} value
 * @return {!proto.ClientSpanProto} returns this
 */
proto.ClientSpanProto.prototype.setBusinessStatusCode = function (value) {
    return jspb.Message.setProto3IntField(this, 15, value);
};

/**
 * optional string to_method = 16;
 * @return {string}
 */
proto.ClientSpanProto.prototype.getToMethod = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 16, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.ClientSpanProto} returns this
 */
proto.ClientSpanProto.prototype.setToMethod = function (value) {
    return jspb.Message.setProto3StringField(this, 16, value);
};

/**
 * optional string to_cluster = 17;
 * @return {string}
 */
proto.ClientSpanProto.prototype.getToCluster = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 17, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.ClientSpanProto} returns this
 */
proto.ClientSpanProto.prototype.setToCluster = function (value) {
    return jspb.Message.setProto3StringField(this, 17, value);
};

/**
 * optional string to_addr = 18;
 * @return {string}
 */
proto.ClientSpanProto.prototype.getToAddr = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 18, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.ClientSpanProto} returns this
 */
proto.ClientSpanProto.prototype.setToAddr = function (value) {
    return jspb.Message.setProto3StringField(this, 18, value);
};

/**
 * optional string to_dc = 19;
 * @return {string}
 */
proto.ClientSpanProto.prototype.getToDc = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 19, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.ClientSpanProto} returns this
 */
proto.ClientSpanProto.prototype.setToDc = function (value) {
    return jspb.Message.setProto3StringField(this, 19, value);
};

/**
 * optional int64 send_size = 20;
 * @return {number}
 */
proto.ClientSpanProto.prototype.getSendSize = function () {
    return /** @type {number} */ (
        jspb.Message.getFieldWithDefault(this, 20, 0)
    );
};

/**
 * @param {number} value
 * @return {!proto.ClientSpanProto} returns this
 */
proto.ClientSpanProto.prototype.setSendSize = function (value) {
    return jspb.Message.setProto3IntField(this, 20, value);
};

/**
 * optional int64 recv_size = 21;
 * @return {number}
 */
proto.ClientSpanProto.prototype.getRecvSize = function () {
    return /** @type {number} */ (
        jspb.Message.getFieldWithDefault(this, 21, 0)
    );
};

/**
 * @param {number} value
 * @return {!proto.ClientSpanProto} returns this
 */
proto.ClientSpanProto.prototype.setRecvSize = function (value) {
    return jspb.Message.setProto3IntField(this, 21, value);
};

/**
 * List of repeated fields within this message type.
 * @private {!Array<number>}
 * @const
 */
proto.EventProto.repeatedFields_ = [3];

if (jspb.Message.GENERATE_TO_OBJECT) {
    /**
     * Creates an object representation of this proto.
     * Field names that are reserved in JavaScript and will be renamed to pb_name.
     * Optional fields that are not set will be set to undefined.
     * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
     * For the list of reserved names please see:
     *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
     * @param {boolean=} opt_includeInstance Deprecated. whether to include the
     *     JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @return {!Object}
     */
    proto.EventProto.prototype.toObject = function (opt_includeInstance) {
        return proto.EventProto.toObject(opt_includeInstance, this);
    };

    /**
     * Static version of the {@see toObject} method.
     * @param {boolean|undefined} includeInstance Deprecated. Whether to include
     *     the JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @param {!proto.EventProto} msg The msg instance to transform.
     * @return {!Object}
     * @suppress {unusedLocalVariables} f is only used for nested messages
     */
    proto.EventProto.toObject = function (includeInstance, msg) {
        var f,
            obj = {
                type: jspb.Message.getFieldWithDefault(msg, 1, ''),
                name: jspb.Message.getFieldWithDefault(msg, 2, ''),
                tagsList: jspb.Message.toObjectList(
                    msg.getTagsList(),
                    proto.Tag.toObject,
                    includeInstance
                ),
                startTime: jspb.Message.getFieldWithDefault(msg, 4, 0),
                data: msg.getData_asB64(),
            };

        if (includeInstance) {
            obj.$jspbMessageInstance = msg;
        }
        return obj;
    };
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.EventProto}
 */
proto.EventProto.deserializeBinary = function (bytes) {
    var reader = new jspb.BinaryReader(bytes);
    var msg = new proto.EventProto();
    return proto.EventProto.deserializeBinaryFromReader(msg, reader);
};

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.EventProto} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.EventProto}
 */
proto.EventProto.deserializeBinaryFromReader = function (msg, reader) {
    while (reader.nextField()) {
        if (reader.isEndGroup()) {
            break;
        }
        var field = reader.getFieldNumber();
        switch (field) {
            case 1:
                var value = /** @type {string} */ (reader.readString());
                msg.setType(value);
                break;
            case 2:
                var value = /** @type {string} */ (reader.readString());
                msg.setName(value);
                break;
            case 3:
                var value = new proto.Tag();
                reader.readMessage(
                    value,
                    proto.Tag.deserializeBinaryFromReader
                );
                msg.addTags(value);
                break;
            case 4:
                var value = /** @type {number} */ (reader.readInt64());
                msg.setStartTime(value);
                break;
            case 5:
                var value = /** @type {!Uint8Array} */ (reader.readBytes());
                msg.setData(value);
                break;
            default:
                reader.skipField();
                break;
        }
    }
    return msg;
};

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.EventProto.prototype.serializeBinary = function () {
    var writer = new jspb.BinaryWriter();
    proto.EventProto.serializeBinaryToWriter(this, writer);
    return writer.getResultBuffer();
};

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.EventProto} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.EventProto.serializeBinaryToWriter = function (message, writer) {
    var f = undefined;
    f = message.getType();
    if (f.length > 0) {
        writer.writeString(1, f);
    }
    f = message.getName();
    if (f.length > 0) {
        writer.writeString(2, f);
    }
    f = message.getTagsList();
    if (f.length > 0) {
        writer.writeRepeatedMessage(3, f, proto.Tag.serializeBinaryToWriter);
    }
    f = message.getStartTime();
    if (f !== 0) {
        writer.writeInt64(4, f);
    }
    f = message.getData_asU8();
    if (f.length > 0) {
        writer.writeBytes(5, f);
    }
};

/**
 * optional string type = 1;
 * @return {string}
 */
proto.EventProto.prototype.getType = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 1, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.EventProto} returns this
 */
proto.EventProto.prototype.setType = function (value) {
    return jspb.Message.setProto3StringField(this, 1, value);
};

/**
 * optional string name = 2;
 * @return {string}
 */
proto.EventProto.prototype.getName = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 2, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.EventProto} returns this
 */
proto.EventProto.prototype.setName = function (value) {
    return jspb.Message.setProto3StringField(this, 2, value);
};

/**
 * repeated Tag tags = 3;
 * @return {!Array<!proto.Tag>}
 */
proto.EventProto.prototype.getTagsList = function () {
    return /** @type{!Array<!proto.Tag>} */ (
        jspb.Message.getRepeatedWrapperField(this, proto.Tag, 3)
    );
};

/**
 * @param {!Array<!proto.Tag>} value
 * @return {!proto.EventProto} returns this
 */
proto.EventProto.prototype.setTagsList = function (value) {
    return jspb.Message.setRepeatedWrapperField(this, 3, value);
};

/**
 * @param {!proto.Tag=} opt_value
 * @param {number=} opt_index
 * @return {!proto.Tag}
 */
proto.EventProto.prototype.addTags = function (opt_value, opt_index) {
    return jspb.Message.addToRepeatedWrapperField(
        this,
        3,
        opt_value,
        proto.Tag,
        opt_index
    );
};

/**
 * Clears the list making it empty but non-null.
 * @return {!proto.EventProto} returns this
 */
proto.EventProto.prototype.clearTagsList = function () {
    return this.setTagsList([]);
};

/**
 * optional int64 start_time = 4;
 * @return {number}
 */
proto.EventProto.prototype.getStartTime = function () {
    return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 4, 0));
};

/**
 * @param {number} value
 * @return {!proto.EventProto} returns this
 */
proto.EventProto.prototype.setStartTime = function (value) {
    return jspb.Message.setProto3IntField(this, 4, value);
};

/**
 * optional bytes data = 5;
 * @return {!(string|Uint8Array)}
 */
proto.EventProto.prototype.getData = function () {
    return /** @type {!(string|Uint8Array)} */ (
        jspb.Message.getFieldWithDefault(this, 5, '')
    );
};

/**
 * optional bytes data = 5;
 * This is a type-conversion wrapper around `getData()`
 * @return {string}
 */
proto.EventProto.prototype.getData_asB64 = function () {
    return /** @type {string} */ (jspb.Message.bytesAsB64(this.getData()));
};

/**
 * optional bytes data = 5;
 * Note that Uint8Array is not supported on all browsers.
 * @see http://caniuse.com/Uint8Array
 * This is a type-conversion wrapper around `getData()`
 * @return {!Uint8Array}
 */
proto.EventProto.prototype.getData_asU8 = function () {
    return /** @type {!Uint8Array} */ (jspb.Message.bytesAsU8(this.getData()));
};

/**
 * @param {!(string|Uint8Array)} value
 * @return {!proto.EventProto} returns this
 */
proto.EventProto.prototype.setData = function (value) {
    return jspb.Message.setProto3BytesField(this, 5, value);
};

/**
 * List of repeated fields within this message type.
 * @private {!Array<number>}
 * @const
 */
proto.MetricProto.repeatedFields_ = [3];

if (jspb.Message.GENERATE_TO_OBJECT) {
    /**
     * Creates an object representation of this proto.
     * Field names that are reserved in JavaScript and will be renamed to pb_name.
     * Optional fields that are not set will be set to undefined.
     * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
     * For the list of reserved names please see:
     *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
     * @param {boolean=} opt_includeInstance Deprecated. whether to include the
     *     JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @return {!Object}
     */
    proto.MetricProto.prototype.toObject = function (opt_includeInstance) {
        return proto.MetricProto.toObject(opt_includeInstance, this);
    };

    /**
     * Static version of the {@see toObject} method.
     * @param {boolean|undefined} includeInstance Deprecated. Whether to include
     *     the JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @param {!proto.MetricProto} msg The msg instance to transform.
     * @return {!Object}
     * @suppress {unusedLocalVariables} f is only used for nested messages
     */
    proto.MetricProto.toObject = function (includeInstance, msg) {
        var f,
            obj = {
                metricName: jspb.Message.getFieldWithDefault(msg, 1, ''),
                metricType: jspb.Message.getFieldWithDefault(msg, 2, ''),
                tagsList: jspb.Message.toObjectList(
                    msg.getTagsList(),
                    proto.Tag.toObject,
                    includeInstance
                ),
                startTime: jspb.Message.getFieldWithDefault(msg, 4, 0),
                value: jspb.Message.getFloatingPointFieldWithDefault(
                    msg,
                    5,
                    0.0
                ),
            };

        if (includeInstance) {
            obj.$jspbMessageInstance = msg;
        }
        return obj;
    };
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.MetricProto}
 */
proto.MetricProto.deserializeBinary = function (bytes) {
    var reader = new jspb.BinaryReader(bytes);
    var msg = new proto.MetricProto();
    return proto.MetricProto.deserializeBinaryFromReader(msg, reader);
};

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.MetricProto} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.MetricProto}
 */
proto.MetricProto.deserializeBinaryFromReader = function (msg, reader) {
    while (reader.nextField()) {
        if (reader.isEndGroup()) {
            break;
        }
        var field = reader.getFieldNumber();
        switch (field) {
            case 1:
                var value = /** @type {string} */ (reader.readString());
                msg.setMetricName(value);
                break;
            case 2:
                var value = /** @type {string} */ (reader.readString());
                msg.setMetricType(value);
                break;
            case 3:
                var value = new proto.Tag();
                reader.readMessage(
                    value,
                    proto.Tag.deserializeBinaryFromReader
                );
                msg.addTags(value);
                break;
            case 4:
                var value = /** @type {number} */ (reader.readInt64());
                msg.setStartTime(value);
                break;
            case 5:
                var value = /** @type {number} */ (reader.readDouble());
                msg.setValue(value);
                break;
            default:
                reader.skipField();
                break;
        }
    }
    return msg;
};

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.MetricProto.prototype.serializeBinary = function () {
    var writer = new jspb.BinaryWriter();
    proto.MetricProto.serializeBinaryToWriter(this, writer);
    return writer.getResultBuffer();
};

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.MetricProto} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.MetricProto.serializeBinaryToWriter = function (message, writer) {
    var f = undefined;
    f = message.getMetricName();
    if (f.length > 0) {
        writer.writeString(1, f);
    }
    f = message.getMetricType();
    if (f.length > 0) {
        writer.writeString(2, f);
    }
    f = message.getTagsList();
    if (f.length > 0) {
        writer.writeRepeatedMessage(3, f, proto.Tag.serializeBinaryToWriter);
    }
    f = message.getStartTime();
    if (f !== 0) {
        writer.writeInt64(4, f);
    }
    f = message.getValue();
    if (f !== 0.0) {
        writer.writeDouble(5, f);
    }
};

/**
 * optional string metric_name = 1;
 * @return {string}
 */
proto.MetricProto.prototype.getMetricName = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 1, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.MetricProto} returns this
 */
proto.MetricProto.prototype.setMetricName = function (value) {
    return jspb.Message.setProto3StringField(this, 1, value);
};

/**
 * optional string metric_type = 2;
 * @return {string}
 */
proto.MetricProto.prototype.getMetricType = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 2, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.MetricProto} returns this
 */
proto.MetricProto.prototype.setMetricType = function (value) {
    return jspb.Message.setProto3StringField(this, 2, value);
};

/**
 * repeated Tag tags = 3;
 * @return {!Array<!proto.Tag>}
 */
proto.MetricProto.prototype.getTagsList = function () {
    return /** @type{!Array<!proto.Tag>} */ (
        jspb.Message.getRepeatedWrapperField(this, proto.Tag, 3)
    );
};

/**
 * @param {!Array<!proto.Tag>} value
 * @return {!proto.MetricProto} returns this
 */
proto.MetricProto.prototype.setTagsList = function (value) {
    return jspb.Message.setRepeatedWrapperField(this, 3, value);
};

/**
 * @param {!proto.Tag=} opt_value
 * @param {number=} opt_index
 * @return {!proto.Tag}
 */
proto.MetricProto.prototype.addTags = function (opt_value, opt_index) {
    return jspb.Message.addToRepeatedWrapperField(
        this,
        3,
        opt_value,
        proto.Tag,
        opt_index
    );
};

/**
 * Clears the list making it empty but non-null.
 * @return {!proto.MetricProto} returns this
 */
proto.MetricProto.prototype.clearTagsList = function () {
    return this.setTagsList([]);
};

/**
 * optional int64 start_time = 4;
 * @return {number}
 */
proto.MetricProto.prototype.getStartTime = function () {
    return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 4, 0));
};

/**
 * @param {number} value
 * @return {!proto.MetricProto} returns this
 */
proto.MetricProto.prototype.setStartTime = function (value) {
    return jspb.Message.setProto3IntField(this, 4, value);
};

/**
 * optional double value = 5;
 * @return {number}
 */
proto.MetricProto.prototype.getValue = function () {
    return /** @type {number} */ (
        jspb.Message.getFloatingPointFieldWithDefault(this, 5, 0.0)
    );
};

/**
 * @param {number} value
 * @return {!proto.MetricProto} returns this
 */
proto.MetricProto.prototype.setValue = function (value) {
    return jspb.Message.setProto3FloatField(this, 5, value);
};

if (jspb.Message.GENERATE_TO_OBJECT) {
    /**
     * Creates an object representation of this proto.
     * Field names that are reserved in JavaScript and will be renamed to pb_name.
     * Optional fields that are not set will be set to undefined.
     * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
     * For the list of reserved names please see:
     *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
     * @param {boolean=} opt_includeInstance Deprecated. whether to include the
     *     JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @return {!Object}
     */
    proto.Tag.prototype.toObject = function (opt_includeInstance) {
        return proto.Tag.toObject(opt_includeInstance, this);
    };

    /**
     * Static version of the {@see toObject} method.
     * @param {boolean|undefined} includeInstance Deprecated. Whether to include
     *     the JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @param {!proto.Tag} msg The msg instance to transform.
     * @return {!Object}
     * @suppress {unusedLocalVariables} f is only used for nested messages
     */
    proto.Tag.toObject = function (includeInstance, msg) {
        var f,
            obj = {
                key: jspb.Message.getFieldWithDefault(msg, 1, ''),
                tagType: jspb.Message.getFieldWithDefault(msg, 2, 0),
                vStr: jspb.Message.getFieldWithDefault(msg, 3, ''),
                vDouble: jspb.Message.getFloatingPointFieldWithDefault(
                    msg,
                    4,
                    0.0
                ),
                vBool: jspb.Message.getBooleanFieldWithDefault(msg, 5, false),
                vLong: jspb.Message.getFieldWithDefault(msg, 6, 0),
                vBytes: msg.getVBytes_asB64(),
            };

        if (includeInstance) {
            obj.$jspbMessageInstance = msg;
        }
        return obj;
    };
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.Tag}
 */
proto.Tag.deserializeBinary = function (bytes) {
    var reader = new jspb.BinaryReader(bytes);
    var msg = new proto.Tag();
    return proto.Tag.deserializeBinaryFromReader(msg, reader);
};

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.Tag} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.Tag}
 */
proto.Tag.deserializeBinaryFromReader = function (msg, reader) {
    while (reader.nextField()) {
        if (reader.isEndGroup()) {
            break;
        }
        var field = reader.getFieldNumber();
        switch (field) {
            case 1:
                var value = /** @type {string} */ (reader.readString());
                msg.setKey(value);
                break;
            case 2:
                var value = /** @type {!proto.TagType} */ (reader.readEnum());
                msg.setTagType(value);
                break;
            case 3:
                var value = /** @type {string} */ (reader.readString());
                msg.setVStr(value);
                break;
            case 4:
                var value = /** @type {number} */ (reader.readDouble());
                msg.setVDouble(value);
                break;
            case 5:
                var value = /** @type {boolean} */ (reader.readBool());
                msg.setVBool(value);
                break;
            case 6:
                var value = /** @type {number} */ (reader.readInt64());
                msg.setVLong(value);
                break;
            case 7:
                var value = /** @type {!Uint8Array} */ (reader.readBytes());
                msg.setVBytes(value);
                break;
            default:
                reader.skipField();
                break;
        }
    }
    return msg;
};

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.Tag.prototype.serializeBinary = function () {
    var writer = new jspb.BinaryWriter();
    proto.Tag.serializeBinaryToWriter(this, writer);
    return writer.getResultBuffer();
};

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.Tag} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.Tag.serializeBinaryToWriter = function (message, writer) {
    var f = undefined;
    f = message.getKey();
    if (f.length > 0) {
        writer.writeString(1, f);
    }
    f = message.getTagType();
    if (f !== 0.0) {
        writer.writeEnum(2, f);
    }
    f = message.getVStr();
    if (f.length > 0) {
        writer.writeString(3, f);
    }
    f = message.getVDouble();
    if (f !== 0.0) {
        writer.writeDouble(4, f);
    }
    f = message.getVBool();
    if (f) {
        writer.writeBool(5, f);
    }
    f = message.getVLong();
    if (f !== 0) {
        writer.writeInt64(6, f);
    }
    f = message.getVBytes_asU8();
    if (f.length > 0) {
        writer.writeBytes(7, f);
    }
};

/**
 * optional string key = 1;
 * @return {string}
 */
proto.Tag.prototype.getKey = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 1, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.Tag} returns this
 */
proto.Tag.prototype.setKey = function (value) {
    return jspb.Message.setProto3StringField(this, 1, value);
};

/**
 * optional TagType tag_type = 2;
 * @return {!proto.TagType}
 */
proto.Tag.prototype.getTagType = function () {
    return /** @type {!proto.TagType} */ (
        jspb.Message.getFieldWithDefault(this, 2, 0)
    );
};

/**
 * @param {!proto.TagType} value
 * @return {!proto.Tag} returns this
 */
proto.Tag.prototype.setTagType = function (value) {
    return jspb.Message.setProto3EnumField(this, 2, value);
};

/**
 * optional string v_str = 3;
 * @return {string}
 */
proto.Tag.prototype.getVStr = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 3, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.Tag} returns this
 */
proto.Tag.prototype.setVStr = function (value) {
    return jspb.Message.setProto3StringField(this, 3, value);
};

/**
 * optional double v_double = 4;
 * @return {number}
 */
proto.Tag.prototype.getVDouble = function () {
    return /** @type {number} */ (
        jspb.Message.getFloatingPointFieldWithDefault(this, 4, 0.0)
    );
};

/**
 * @param {number} value
 * @return {!proto.Tag} returns this
 */
proto.Tag.prototype.setVDouble = function (value) {
    return jspb.Message.setProto3FloatField(this, 4, value);
};

/**
 * optional bool v_bool = 5;
 * @return {boolean}
 */
proto.Tag.prototype.getVBool = function () {
    return /** @type {boolean} */ (
        jspb.Message.getBooleanFieldWithDefault(this, 5, false)
    );
};

/**
 * @param {boolean} value
 * @return {!proto.Tag} returns this
 */
proto.Tag.prototype.setVBool = function (value) {
    return jspb.Message.setProto3BooleanField(this, 5, value);
};

/**
 * optional int64 v_long = 6;
 * @return {number}
 */
proto.Tag.prototype.getVLong = function () {
    return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 6, 0));
};

/**
 * @param {number} value
 * @return {!proto.Tag} returns this
 */
proto.Tag.prototype.setVLong = function (value) {
    return jspb.Message.setProto3IntField(this, 6, value);
};

/**
 * optional bytes v_bytes = 7;
 * @return {!(string|Uint8Array)}
 */
proto.Tag.prototype.getVBytes = function () {
    return /** @type {!(string|Uint8Array)} */ (
        jspb.Message.getFieldWithDefault(this, 7, '')
    );
};

/**
 * optional bytes v_bytes = 7;
 * This is a type-conversion wrapper around `getVBytes()`
 * @return {string}
 */
proto.Tag.prototype.getVBytes_asB64 = function () {
    return /** @type {string} */ (jspb.Message.bytesAsB64(this.getVBytes()));
};

/**
 * optional bytes v_bytes = 7;
 * Note that Uint8Array is not supported on all browsers.
 * @see http://caniuse.com/Uint8Array
 * This is a type-conversion wrapper around `getVBytes()`
 * @return {!Uint8Array}
 */
proto.Tag.prototype.getVBytes_asU8 = function () {
    return /** @type {!Uint8Array} */ (
        jspb.Message.bytesAsU8(this.getVBytes())
    );
};

/**
 * @param {!(string|Uint8Array)} value
 * @return {!proto.Tag} returns this
 */
proto.Tag.prototype.setVBytes = function (value) {
    return jspb.Message.setProto3BytesField(this, 7, value);
};

if (jspb.Message.GENERATE_TO_OBJECT) {
    /**
     * Creates an object representation of this proto.
     * Field names that are reserved in JavaScript and will be renamed to pb_name.
     * Optional fields that are not set will be set to undefined.
     * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
     * For the list of reserved names please see:
     *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
     * @param {boolean=} opt_includeInstance Deprecated. whether to include the
     *     JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @return {!Object}
     */
    proto.TraceID.prototype.toObject = function (opt_includeInstance) {
        return proto.TraceID.toObject(opt_includeInstance, this);
    };

    /**
     * Static version of the {@see toObject} method.
     * @param {boolean|undefined} includeInstance Deprecated. Whether to include
     *     the JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @param {!proto.TraceID} msg The msg instance to transform.
     * @return {!Object}
     * @suppress {unusedLocalVariables} f is only used for nested messages
     */
    proto.TraceID.toObject = function (includeInstance, msg) {
        var f,
            obj = {
                low: jspb.Message.getFieldWithDefault(msg, 1, '0'),
                high: jspb.Message.getFieldWithDefault(msg, 2, '0'),
            };

        if (includeInstance) {
            obj.$jspbMessageInstance = msg;
        }
        return obj;
    };
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.TraceID}
 */
proto.TraceID.deserializeBinary = function (bytes) {
    var reader = new jspb.BinaryReader(bytes);
    var msg = new proto.TraceID();
    return proto.TraceID.deserializeBinaryFromReader(msg, reader);
};

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.TraceID} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.TraceID}
 */
proto.TraceID.deserializeBinaryFromReader = function (msg, reader) {
    while (reader.nextField()) {
        if (reader.isEndGroup()) {
            break;
        }
        var field = reader.getFieldNumber();
        switch (field) {
            case 1:
                var value = /** @type {string} */ (reader.readUint64String());
                msg.setLow(value);
                break;
            case 2:
                var value = /** @type {string} */ (reader.readUint64String());
                msg.setHigh(value);
                break;
            default:
                reader.skipField();
                break;
        }
    }
    return msg;
};

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.TraceID.prototype.serializeBinary = function () {
    var writer = new jspb.BinaryWriter();
    proto.TraceID.serializeBinaryToWriter(this, writer);
    return writer.getResultBuffer();
};

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.TraceID} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.TraceID.serializeBinaryToWriter = function (message, writer) {
    var f = undefined;
    f = message.getLow();
    if (parseInt(f, 10) !== 0) {
        writer.writeUint64String(1, f);
    }
    f = message.getHigh();
    if (parseInt(f, 10) !== 0) {
        writer.writeUint64String(2, f);
    }
};

/**
 * optional uint64 low = 1;
 * @return {string}
 */
proto.TraceID.prototype.getLow = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 1, '0')
    );
};

/**
 * @param {string} value
 * @return {!proto.TraceID} returns this
 */
proto.TraceID.prototype.setLow = function (value) {
    return jspb.Message.setProto3StringIntField(this, 1, value);
};

/**
 * optional uint64 high = 2;
 * @return {string}
 */
proto.TraceID.prototype.getHigh = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 2, '0')
    );
};

/**
 * @param {string} value
 * @return {!proto.TraceID} returns this
 */
proto.TraceID.prototype.setHigh = function (value) {
    return jspb.Message.setProto3StringIntField(this, 2, value);
};

if (jspb.Message.GENERATE_TO_OBJECT) {
    /**
     * Creates an object representation of this proto.
     * Field names that are reserved in JavaScript and will be renamed to pb_name.
     * Optional fields that are not set will be set to undefined.
     * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
     * For the list of reserved names please see:
     *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
     * @param {boolean=} opt_includeInstance Deprecated. whether to include the
     *     JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @return {!Object}
     */
    proto.SpanRef.prototype.toObject = function (opt_includeInstance) {
        return proto.SpanRef.toObject(opt_includeInstance, this);
    };

    /**
     * Static version of the {@see toObject} method.
     * @param {boolean|undefined} includeInstance Deprecated. Whether to include
     *     the JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @param {!proto.SpanRef} msg The msg instance to transform.
     * @return {!Object}
     * @suppress {unusedLocalVariables} f is only used for nested messages
     */
    proto.SpanRef.toObject = function (includeInstance, msg) {
        var f,
            obj = {
                traceId:
                    (f = msg.getTraceId()) &&
                    proto.TraceID.toObject(includeInstance, f),
                parentSpanId: jspb.Message.getFieldWithDefault(msg, 2, '0'),
                refType: jspb.Message.getFieldWithDefault(msg, 3, 0),
            };

        if (includeInstance) {
            obj.$jspbMessageInstance = msg;
        }
        return obj;
    };
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.SpanRef}
 */
proto.SpanRef.deserializeBinary = function (bytes) {
    var reader = new jspb.BinaryReader(bytes);
    var msg = new proto.SpanRef();
    return proto.SpanRef.deserializeBinaryFromReader(msg, reader);
};

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.SpanRef} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.SpanRef}
 */
proto.SpanRef.deserializeBinaryFromReader = function (msg, reader) {
    while (reader.nextField()) {
        if (reader.isEndGroup()) {
            break;
        }
        var field = reader.getFieldNumber();
        switch (field) {
            case 1:
                var value = new proto.TraceID();
                reader.readMessage(
                    value,
                    proto.TraceID.deserializeBinaryFromReader
                );
                msg.setTraceId(value);
                break;
            case 2:
                var value = /** @type {string} */ (reader.readUint64String());
                msg.setParentSpanId(value);
                break;
            case 3:
                var value = /** @type {number} */ (reader.readInt32());
                msg.setRefType(value);
                break;
            default:
                reader.skipField();
                break;
        }
    }
    return msg;
};

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.SpanRef.prototype.serializeBinary = function () {
    var writer = new jspb.BinaryWriter();
    proto.SpanRef.serializeBinaryToWriter(this, writer);
    return writer.getResultBuffer();
};

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.SpanRef} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.SpanRef.serializeBinaryToWriter = function (message, writer) {
    var f = undefined;
    f = message.getTraceId();
    if (f != null) {
        writer.writeMessage(1, f, proto.TraceID.serializeBinaryToWriter);
    }
    f = message.getParentSpanId();
    if (parseInt(f, 10) !== 0) {
        writer.writeUint64String(2, f);
    }
    f = message.getRefType();
    if (f !== 0) {
        writer.writeInt32(3, f);
    }
};

/**
 * optional TraceID trace_id = 1;
 * @return {?proto.TraceID}
 */
proto.SpanRef.prototype.getTraceId = function () {
    return /** @type{?proto.TraceID} */ (
        jspb.Message.getWrapperField(this, proto.TraceID, 1)
    );
};

/**
 * @param {?proto.TraceID|undefined} value
 * @return {!proto.SpanRef} returns this
 */
proto.SpanRef.prototype.setTraceId = function (value) {
    return jspb.Message.setWrapperField(this, 1, value);
};

/**
 * Clears the message field making it undefined.
 * @return {!proto.SpanRef} returns this
 */
proto.SpanRef.prototype.clearTraceId = function () {
    return this.setTraceId(undefined);
};

/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.SpanRef.prototype.hasTraceId = function () {
    return jspb.Message.getField(this, 1) != null;
};

/**
 * optional uint64 parent_span_id = 2;
 * @return {string}
 */
proto.SpanRef.prototype.getParentSpanId = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 2, '0')
    );
};

/**
 * @param {string} value
 * @return {!proto.SpanRef} returns this
 */
proto.SpanRef.prototype.setParentSpanId = function (value) {
    return jspb.Message.setProto3StringIntField(this, 2, value);
};

/**
 * optional int32 ref_type = 3;
 * @return {number}
 */
proto.SpanRef.prototype.getRefType = function () {
    return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 3, 0));
};

/**
 * @param {number} value
 * @return {!proto.SpanRef} returns this
 */
proto.SpanRef.prototype.setRefType = function (value) {
    return jspb.Message.setProto3IntField(this, 3, value);
};

/**
 * @enum {number}
 */
proto.TagType = {
    STRING: 0,
    DOUBLE: 1,
    BOOL: 2,
    LONG: 3,
    BYTES: 4,
};

goog.object.extend(exports, proto);
