// @universe-design/react-ai/lib-commonjs/card/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AICard = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _reactUtilities = require("@universe-design/react-utilities");
var _react2 = require("@universe-design/react");
var _clsx = _interopRequireDefault(require("clsx"));
require("./styles/index.less");
/**
 * @public
 */

/**
 * @public
 */
const AICard = exports.AICard = /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
  const {
    children,
    className,
    processing = false,
    shadowed = true,
    noHoverState = false,
    ...restProps
  } = props;
  // aiCard 的尺寸
  const [{
    width,
    height
  }, setAiCardSize] = (0, _react.useState)({
    width: 0,
    height: 0
  });

  // 获取 aiCard 的尺寸
  const aiCardRef = (0, _react2.useResizeObserver)(entry => {
    const {
      contentRect: {
        width: _width,
        height: _height
      }
    } = entry;
    setAiCardSize({
      width: _width,
      height: _height
    });
  });

  // 求容器对角线的长度
  const maskR = (0, _react.useMemo)(() => Math.hypot(width, height), [height, width]);
  const mergedRef = (0, _react.useMemo)(() => (0, _reactUtilities.composeRefs)(aiCardRef, ref), [aiCardRef, ref]);
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
    ref: mergedRef,
    className: (0, _clsx.default)('udx__ai__card', {
      ['udx__ai__card--default']: !processing,
      ['udx__ai__card--noshadows']: !shadowed,
      ['udx__ai__card--no-hover-state']: noHoverState
    }, className)
  }, restProps), processing && /*#__PURE__*/_react.default.createElement("div", {
    className: "udx__ai__card__mask",
    style: {
      width: maskR,
      height: maskR,
      borderRadius: '50%',
      top: -(maskR - height) / 2,
      left: -(maskR - width) / 2
    }
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)('udx__ai__card__content', processing && 'udx__ai__card__content--processing')
  }, children));
});
