// @universe-design/react/src/cascader/stories/homeless.stories.tsx
import React, { useState } from 'react';
import type { Meta } from '@storybook/react';
import { Button, Cascader, Space, Switch } from '../..';
import type { TreeConfig, TreeNodeConfig } from '../..';
import { __SCREENSHOT__ } from '../../util';
import Cascade from '../cascade';
import type { CascadeRef } from '../interface';

export default {
  title: 'Components/Cascader/临时示例',
  component: Cascader,
} satisfies Meta;

const options: TreeNodeConfig[] = [
  {
    value: 'zhejiang',
    label: 'Zhejiang',
    children: [
      {
        value: 'hangzhou',
        label: 'Hangzhou',
        children: [
          {
            value: 'xihu',
            label: 'West Lake',
          },
        ],
      },
    ],
  },
  {
    value: 'jiangsu',
    label: 'Jiangsu',
    children: [
      {
        value: 'nanjing',
        label: 'Nanjing',
        children: [...new Array(100)].map((_, index) => ({ value: String(index) } satisfies TreeNodeConfig)),
      },
    ],
  },
  {
    value: 'beijing',
    label: 'Beijing',
    disabled: true,
    children: [
      {
        value: 'haidian',
        label: 'Haidian',
        disabled: true,
      },
    ],
  },
];

const loadableTreeDate: TreeConfig = [
  { label: 'Expand to load', value: '0' },
  { label: 'Expand to load', value: '1' },
  { label: 'Tree Node', value: '2', isLeaf: true },
];

function updateTreeData(list: TreeConfig, key: React.Key, children: TreeConfig): TreeConfig {
  return list.map(node => {
    if (node.value === key) {
      return {
        ...node,
        children,
      };
    } else if (node.children) {
      return {
        ...node,
        children: updateTreeData(node.children, key, children),
      };
    }
    return node;
  });
}

// 递归搜索树节点
const searchTreeNodes = (nodes: TreeNodeConfig[], keyword: string): TreeNodeConfig[] =>
  nodes.reduce<TreeNodeConfig[]>((filtered, node) => {
    const matchesKeyword = (node.label as string).toLowerCase().includes(keyword.toLowerCase());
    const children = node.children ? searchTreeNodes(node.children, keyword) : undefined;

    if (matchesKeyword || children?.length) {
      filtered.push({
        ...node,
        children: children || node.children,
      });
    }
    return filtered;
  }, []);

// 模拟后端返回的数据
const mockData: TreeNodeConfig[] = [
  {
    label: 'Expand to load',
    value: '0',
    children: [
      {
        value: '0-0',
        label: '0-0',
        children: [
          { value: '0-0-0', label: '0-0-0', isLeaf: true },
          { value: '0-0-1', label: '0-0-1', isLeaf: true },
        ],
      },
      {
        value: '0-1',
        label: '0-1',
        children: [
          { value: '0-1-0', label: '0-1-0', isLeaf: true },
          { value: '0-1-1', label: '0-1-1', isLeaf: true },
        ],
      },
    ],
  },
  {
    label: 'Expand to load',
    value: '1',
    children: [
      {
        value: '1-0',
        label: '1-0',
        children: [
          { value: '1-0-0', label: '1-0-0', isLeaf: true },
          { value: '1-0-1', label: '1-0-1', isLeaf: true },
        ],
      },
      {
        value: '1-1',
        label: '1-1',
        children: [
          { value: '1-1-0', label: '1-1-0', isLeaf: true },
          { value: '1-1-1', label: '1-1-1', isLeaf: true },
        ],
      },
    ],
  },
];

// 模拟后端搜索接口
const mockSearchFromServer = async (keyword: string): Promise<TreeNodeConfig[]> => {
  // 模拟API延迟
  await new Promise(resolve => setTimeout(resolve, 300));
  return searchTreeNodes(mockData, keyword);
};

const findNodeByKey = (nodes: TreeNodeConfig[], key: string): TreeNodeConfig | null => {
  for (const node of nodes) {
    if (node.value === key) {
      return node;
    }
    if (node.children) {
      const found = findNodeByKey(node.children, key);
      if (found) {
        return found;
      }
    }
  }
  return null;
};

const mockLoadData = async (key: string): Promise<TreeNodeConfig[]> => {
  // 模拟API延迟
  await new Promise(resolve => setTimeout(resolve, 300));

  // 查找对应的节点
  const targetNode = findNodeByKey(mockData, key);

  // 如果找到节点并且有子节点，返回子节点
  if (targetNode?.children) {
    return targetNode.children;
  }

  // 如果没有找到节点或节点没有子节点，返回空数组
  return [];
};

// 合并搜索结果到现有树结构
const mergeSearchResultToTree = (originalTree: TreeConfig, searchResults: TreeNodeConfig[]): TreeConfig => {
  const newTree = [...originalTree];

  searchResults.forEach(result => {
    const existingNodeIndex = newTree.findIndex(node => node.value === result.value);

    if (existingNodeIndex === -1) {
      // 如果节点不存在，直接添加
      newTree.push(result);
    } else {
      // 如果节点存在且有子节点，递归合并子节点
      const existingNode = newTree[existingNodeIndex];
      if (result.children && result.children.length > 0) {
        newTree[existingNodeIndex] = {
          ...existingNode,
          children: existingNode.children
            ? mergeSearchResultToTree(existingNode.children, result.children)
            : result.children,
        };
      }
    }
  });

  return newTree;
};

export const TriggerElement = () => (
  <>
    <Cascader style={{ width: 300 }} options={options} triggerElement={<span>123</span>} />
    <br />
    <br />
    <Cascader
      open={true}
      style={{ width: 300 }}
      options={options}
      triggerElement={({ valueSet }) => <Button>Places: {[...valueSet].join(', ')}</Button>}
    />
  </>
);

export const Search = () => (
  <Cascader
    open
    style={{ width: 300 }}
    options={options}
    onChange={(value, meta) => {
      console.log(value, meta);
    }}
    showSearch
    filterResultLimit={false}
    placeholder="Please select"
    changeOnSelect
    filterResultRenderer={paths => paths.map(node => node.label || node.value).join('|')}
  />
);

export const Dynamic = () => {
  // It's just a simple demo. You can use tree map to optimize update perf.

  const [treeData, setTreeData] = useState(loadableTreeDate);
  const [checkable, setCheckable] = useState(false);
  const [loading, setLoading] = useState(false);

  return (
    <Space>
      <p>
        <Switch checked={checkable} onChange={e => setCheckable(e)} title="ddd" />
        <span>Checkable </span>
      </p>
      <Cascader
        mode="multiple"
        checkable={checkable}
        options={treeData}
        loadData={({ key, children }) => {
          console.log(key, children);
          return new Promise<void>(resolve => {
            if (children) {
              resolve();
              return;
            }
            setTimeout(() => {
              setTreeData(origin =>
                updateTreeData(origin, key, [
                  { label: `${key}-0`, value: `${key}-0` },
                  { label: `${key}-1`, value: `${key}-1` },
                ]),
              );

              resolve();
            }, 1000);
          });
        }}
        onSelect={console.log}
        style={{ width: 300 }}
      />
      <Cascader
        mode="multiple"
        checkable
        style={{ width: 300 }}
        options={treeData}
        loading={loading}
        loadData={({ key }) =>
          new Promise<void>(resolve => {
            mockLoadData(key).then(children => {
              setTreeData(origin => updateTreeData(origin, key, children));
              resolve();
            });
          })
        }
        expandTrigger="hover"
        showSearch
        onSearch={async (inputValue, { action }) => {
          if (action === 1 && inputValue) {
            setLoading(true);
            try {
              const searchResults = await mockSearchFromServer(inputValue);
              const mergedTreeData = mergeSearchResultToTree(treeData, searchResults);
              setTreeData(mergedTreeData);
            } catch (error) {
              console.error('搜索失败:', error);
            } finally {
              setLoading(false);
            }
          }
        }}
        defaultOpen={__SCREENSHOT__}
      />
    </Space>
  );
};

export const InnerCascade = () => {
  const [treeData, setTreeData] = useState(loadableTreeDate);
  const cascadeRef = React.useRef<CascadeRef>(null);
  return (
    <Space direction="vertical">
      <Button onClick={() => cascadeRef.current?.scrollIntoView('33', { offset: 100 })}>Scrollto </Button>
      <Cascade options={options} ref={cascadeRef} multiple />
      <Cascade options={options} expandTrigger="hover" />
      <Cascade
        options={treeData}
        loadData={({ key, children }) => {
          console.log(key, children);
          return new Promise<void>(resolve => {
            if (children) {
              resolve();
              return;
            }
            setTimeout(() => {
              setTreeData(origin =>
                updateTreeData(origin, key, [
                  { label: `${key}-0`, value: `${key}-0`, isLeaf: true },
                  { label: `${key}-1`, value: `${key}-1`, isLeaf: true },
                ]),
              );
              resolve();
            }, 1000);
          });
        }}
      />
    </Space>
  );
};

export const WithRepeatKeys = () => {
  const optionsWithRepeatKeys = [
    {
      label: 'Parent 1',
      value: '0-0',
      children: [
        {
          label: 'Parent 1-1',
          value: '0-0',
        },
        {
          label: 'Parent 1-2',
          value: '0-0',
        },
      ],
    },
    {
      label: 'Parent 2',
      value: 'parent-2',
    },
  ];
  return <Cascader options={optionsWithRepeatKeys} />;
};
