// @universe-design/react/src/cascader/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Basic from '../demo/0.basic';
import C12CustomDropdown from '../demo/12.custom-dropdown';
import C13TriggerElement from '../demo/13.trigger-element';
import C14Dropdown from '../demo/14.dropdown';
import C15ShowCheckedStrategy from '../demo/15.showCheckedStrategy';
import C16FullHeight from '../demo/16.full-height';
import C2Disabled from '../demo/2.disabled';
import C3Multiple from '../demo/3.multiple';
import C4CascadedLabel from '../demo/4.cascaded-label';
import C5ChangeOnSelect from '../demo/5.change-on-select';
import C6DropdownLevelWidth from '../demo/6.dropdown_level_width';
import C7DisplayRenderer from '../demo/7.display-renderer';
import C8Dynamic from '../demo/8.dynamic';
if (!(C0Basic as Story).storyName) {
  (C0Basic as Story).storyName = '0. 基础用法';
  (C0Basic as Story).parameters = {
    docs: {
      description: {
        story: '简单的用法。省市区级联。\n\n',
      },
    },
  };
}
if (!(C12CustomDropdown as Story).storyName) {
  (C12CustomDropdown as Story).storyName = '12. 扩展菜单';
  (C12CustomDropdown as Story).parameters = {
    docs: {
      description: {
        story: '通过`dropdownRender`对下拉菜单进行自由扩展。\n\n',
      },
    },
  };
}
if (!(C13TriggerElement as Story).storyName) {
  (C13TriggerElement as Story).storyName = '13. 自定义触发元素';
  (C13TriggerElement as Story).parameters = {
    docs: {
      description: {
        story: '通过`triggerElement` 替换 `Input` 作为触发元素。\n\n',
      },
    },
  };
}
if (!(C14Dropdown as Story).storyName) {
  (C14Dropdown as Story).storyName = '13. 将 Cascader 放到 popup 里';
  (C14Dropdown as Story).parameters = {
    docs: {
      description: {
        story: '可以利用 `dropdownRender` 实现更复杂的需求。\n\n',
      },
    },
  };
}
if (!(C15ShowCheckedStrategy as Story).storyName) {
  (C15ShowCheckedStrategy as Story).storyName = '15. 选中项回填的方式';
  (C15ShowCheckedStrategy as Story).parameters = {
    docs: {
      description: {
        story:
          '可以通过 `showCheckedStrategy` API 定义选中项回填的方式。当取值为 `Cascader.SHOW_PARENT` 时，如果父节点下所有子节点都选中，只显示父节点；当取值为`Cascader.SHOW_CHILD` 时，如果父节点下所有子节点都选中，只显示子节点（默认行为）。\n\n',
      },
    },
  };
}
if (!(C16FullHeight as Story).storyName) {
  (C16FullHeight as Story).storyName = '16. 下拉菜单自适应高度';
  (C16FullHeight as Story).parameters = {
    docs: {
      description: {
        story: '通过设置 `fullHeight` 使下拉菜单自适应高度。\n\n',
      },
    },
  };
}
if (!(C2Disabled as Story).storyName) {
  (C2Disabled as Story).storyName = '2. 禁用选项';
  (C2Disabled as Story).parameters = {
    docs: {
      description: {
        story: '通过指定`options`里的`disabled`字段，禁用该选项。\n\n',
      },
    },
  };
}
if (!(C3Multiple as Story).storyName) {
  (C3Multiple as Story).storyName = '3. 多选';
  (C3Multiple as Story).parameters = {
    docs: {
      description: {
        story: '多选级联选择\n\n',
      },
    },
  };
}
if (!(C4CascadedLabel as Story).storyName) {
  (C4CascadedLabel as Story).storyName = '4. 选中项显示为级联状态';
  (C4CascadedLabel as Story).parameters = {
    docs: {
      description: {
        story: '可以使用 `fullPathBackfillOption` 属性控制回填到选择器中的选项是否为级联。\n\n',
      },
    },
  };
}
if (!(C5ChangeOnSelect as Story).storyName) {
  (C5ChangeOnSelect as Story).storyName = '5. 选择即改变';
  (C5ChangeOnSelect as Story).parameters = {
    docs: {
      description: {
        story: '这种交互允许只选中父级选项。\n\n',
      },
    },
  };
}
if (!(C6DropdownLevelWidth as Story).storyName) {
  (C6DropdownLevelWidth as Story).storyName = '6. 自定义宽度';
  (C6DropdownLevelWidth as Story).parameters = {
    docs: {
      description: {
        story: '可以通过`dropdownLevelWidth`属性改变下拉菜单每一级的宽度。\n\n',
      },
    },
  };
}
if (!(C7DisplayRenderer as Story).storyName) {
  (C7DisplayRenderer as Story).storyName = '7. 自定义回填内容';
  (C7DisplayRenderer as Story).parameters = {
    docs: {
      description: {
        story: '可以通过 `optionLabelProp` 属性自定义选中后的回填内容。\n\n',
      },
    },
  };
}
if (!(C8Dynamic as Story).storyName) {
  (C8Dynamic as Story).storyName = '8. 异步加载数据';
  (C8Dynamic as Story).parameters = {
    docs: {
      description: {
        story:
          '可以通过 `loadData` 属性并配合节点的 `isLeaf` 属性来实现异步加载数据。除了 `isLeaf` 设置为 true 的节点，其他节点初次点击时都会调用 loadData 加载数据。你也可以通过控制 `loadedKeys` 来实现点击刷新、单独某些节点不请求数据。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Cascader/官网示例',
  tags: ['autodocs'],
};

export {
  C0Basic,
  C12CustomDropdown,
  C13TriggerElement,
  C14Dropdown,
  C15ShowCheckedStrategy,
  C16FullHeight,
  C2Disabled,
  C3Multiple,
  C4CascadedLabel,
  C5ChangeOnSelect,
  C6DropdownLevelWidth,
  C7DisplayRenderer,
  C8Dynamic,
};
