// @universe-design/react/lib/upload/Dragger.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.UploadDragArea = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _FileAddColorful = _interopRequireDefault(require("@universe-design/icons-react/FileAddColorful"));
var _FileAddDisableColorful = _interopRequireDefault(require("@universe-design/icons-react/FileAddDisableColorful"));
var _Upload = require("./Upload");
var _UploadContext = require("./hooks/UploadContext");
var _sx = require("./sx");
var _forwardRef = require("../util/forwardRef");
const UploadDragArea = exports.UploadDragArea = (0, _forwardRef.forwardHTMLRef)((props, ref) => {
  const {
    dragIcon: dragIconProp,
    dragDescription,
    dragTitle,
    className,
    onMouseEnter: onMouseEnterProp,
    onMouseLeave: onMouseLeaveProp,
    onDragLeave: onDragLeaveProp,
    onDragOver: onDragOverProp,
    ...resetProps
  } = props;
  const {
    rootPrefixCls,
    disabled
  } = (0, _UploadContext.useUploadContext)();
  const [isActive, setIsActive] = (0, _react.useState)(false);
  const {
    sxDisabledText,
    sxDragArea,
    sxDragAreaActive,
    sxDragAreaDescription,
    sxDragAreaDisabled,
    sxDragAreaIcon,
    sxDragAreaTitle
  } = (0, _sx.useDragAreaSx)(rootPrefixCls);
  const classNames = (0, _clsx.default)([className, (0, _stylex.default)(sxDragArea, isActive && sxDragAreaActive, disabled && sxDragAreaDisabled)]);
  const dragIcon = dragIconProp !== null && dragIconProp !== void 0 ? dragIconProp : disabledArg => disabledArg ? /*#__PURE__*/_react.default.createElement(_FileAddDisableColorful.default, null) : /*#__PURE__*/_react.default.createElement(_FileAddColorful.default, null);
  const icon = typeof dragIcon === 'function' ? dragIcon(disabled) : dragIcon;
  const events = _react.default.useMemo(() => {
    if (disabled) {
      return {};
    }
    const active = () => setIsActive(true);
    const inactive = () => setIsActive(false);
    return {
      onMouseEnter: e => {
        onMouseEnterProp === null || onMouseEnterProp === void 0 ? void 0 : onMouseEnterProp(e);
        active();
      },
      onMouseLeave: e => {
        onMouseLeaveProp === null || onMouseLeaveProp === void 0 ? void 0 : onMouseLeaveProp(e);
        inactive();
      },
      onDragLeave: e => {
        onDragLeaveProp === null || onDragLeaveProp === void 0 ? void 0 : onDragLeaveProp(e);
        inactive();
      },
      onDragOver: e => {
        onDragOverProp === null || onDragOverProp === void 0 ? void 0 : onDragOverProp(e);
        active();
      }
    };
  }, [disabled, onDragLeaveProp, onDragOverProp, onMouseEnterProp, onMouseLeaveProp]);
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
    className: classNames,
    ref: ref
  }, resetProps, events), /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _stylex.default)(sxDragAreaIcon)
  }, icon), /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _stylex.default)(sxDragAreaTitle, disabled && sxDisabledText)
  }, dragTitle), /*#__PURE__*/_react.default.createElement("p", {
    className: (0, _clsx.default)((0, _stylex.default)(sxDragAreaDescription, disabled && sxDisabledText))
  }, dragDescription));
});
const InternalDragger = (_ref, ref) => {
  let {
    dragAreaConfig,
    ...restProps
  } = _ref;
  return /*#__PURE__*/_react.default.createElement(_Upload.Upload, (0, _extends2.default)({
    ref: ref
  }, restProps, {
    type: "drag"
  }), dragAreaConfig ? /*#__PURE__*/_react.default.createElement(UploadDragArea, dragAreaConfig) : restProps.children);
};
const Dragger = /*#__PURE__*/_react.default.forwardRef(InternalDragger);
if (process.env.NODE_ENV !== "production") {
  Dragger.displayName = 'Dragger';
}
var _default = exports.default = Dragger;
