// @universe-design/react/lib/upload/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CardFile", {
  enumerable: true,
  get: function () {
    return _uploadFile.CardFile;
  }
});
Object.defineProperty(exports, "ComposableUpload", {
  enumerable: true,
  get: function () {
    return _ComposableUpload.ComposableUpload;
  }
});
Object.defineProperty(exports, "PictureFile", {
  enumerable: true,
  get: function () {
    return _uploadFile.PictureFile;
  }
});
Object.defineProperty(exports, "PictureUploadTrigger", {
  enumerable: true,
  get: function () {
    return _PictureUploadTrigger.PictureUploadTrigger;
  }
});
Object.defineProperty(exports, "TextFile", {
  enumerable: true,
  get: function () {
    return _uploadFile.TextFile;
  }
});
Object.defineProperty(exports, "TextFileAction", {
  enumerable: true,
  get: function () {
    return _uploadFile.TextFileAction;
  }
});
exports.Upload = void 0;
Object.defineProperty(exports, "UploadCardFile", {
  enumerable: true,
  get: function () {
    return _uploadFile.UploadCardFile;
  }
});
Object.defineProperty(exports, "UploadCardFileList", {
  enumerable: true,
  get: function () {
    return _uploadList.UploadCardFileList;
  }
});
Object.defineProperty(exports, "UploadDragArea", {
  enumerable: true,
  get: function () {
    return _Dragger.UploadDragArea;
  }
});
Object.defineProperty(exports, "UploadPictureFile", {
  enumerable: true,
  get: function () {
    return _uploadFile.UploadPictureFile;
  }
});
Object.defineProperty(exports, "UploadPictureFileList", {
  enumerable: true,
  get: function () {
    return _uploadList.UploadPictureFileList;
  }
});
Object.defineProperty(exports, "UploadSimpleTextFile", {
  enumerable: true,
  get: function () {
    return _uploadFile.UploadSimpleTextFile;
  }
});
Object.defineProperty(exports, "UploadSimpleTextFileList", {
  enumerable: true,
  get: function () {
    return _uploadList.UploadSimpleTextFileList;
  }
});
Object.defineProperty(exports, "UploadTextFile", {
  enumerable: true,
  get: function () {
    return _uploadFile.UploadTextFile;
  }
});
Object.defineProperty(exports, "UploadTextFileList", {
  enumerable: true,
  get: function () {
    return _uploadList.UploadTextFileList;
  }
});
Object.defineProperty(exports, "UploadTrigger", {
  enumerable: true,
  get: function () {
    return _UploadTrigger.UploadTrigger;
  }
});
exports.default = void 0;
Object.defineProperty(exports, "useUploadContext", {
  enumerable: true,
  get: function () {
    return _UploadContext.useUploadContext;
  }
});
var _Upload = require("./Upload");
var _Dragger = _interopRequireWildcard(require("./Dragger"));
var _UploadTrigger = require("./UploadTrigger");
var _UploadContext = require("./hooks/UploadContext");
var _PictureUploadTrigger = require("./uploadList/PictureUploadTrigger");
var _uploadFile = require("./uploadFile");
var _uploadList = require("./uploadList");
var _ComposableUpload = require("./ComposableUpload");
// uploadFile

// uploadList

const Upload = exports.Upload = _Upload.Upload;
Upload.Dragger = _Dragger.default;
var _default = exports.default = Upload;
