// @universe-design/react/lib/upload/sx.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useDragAreaSx = void 0;
var _unha2if = _interopRequireDefault(require("./../.stylex/unha2if"));
var _u16q6d1w = _interopRequireDefault(require("./../.stylex/u16q6d1w"));
var _u22yt5q = _interopRequireDefault(require("./../.stylex/u22yt5q"));
var _u1kxgv9m = _interopRequireDefault(require("./../.stylex/u1kxgv9m"));
var _unbze3l = _interopRequireDefault(require("./../.stylex/unbze3l"));
var _u10d06an = _interopRequireDefault(require("./../.stylex/u10d06an"));
var _u1w6ojc = _interopRequireDefault(require("./../.stylex/u1w6ojc6"));
var _u1dhzo8t = _interopRequireDefault(require("./../.stylex/u1dhzo8t"));
var _u19bueuo = _interopRequireDefault(require("./../.stylex/u19bueuo"));
var _uxmlixj = _interopRequireDefault(require("./../.stylex/uxmlixj"));
var _u8ckwbm = _interopRequireDefault(require("./../.stylex/u8ckwbm"));
var _uv7gtgj = _interopRequireDefault(require("./../.stylex/uv7gtgj"));
var _u1hkmcjz = _interopRequireDefault(require("./../.stylex/u1hkmcjz"));
var _u1hqmxtl = _interopRequireDefault(require("./../.stylex/u1hqmxtl"));
var _u1odtry = _interopRequireDefault(require("./../.stylex/u1odtry2"));
var _ug8e45j = _interopRequireDefault(require("./../.stylex/ug8e45j"));
var _u1vjduc = _interopRequireDefault(require("./../.stylex/u1vjduc2"));
var _u12o6wmt = _interopRequireDefault(require("./../.stylex/u12o6wmt"));
var _u1cpcpb = _interopRequireDefault(require("./../.stylex/u1cpcpb5"));
var _u10qn3go = _interopRequireDefault(require("./../.stylex/u10qn3go"));
var _uflkg = _interopRequireDefault(require("./../.stylex/uflkg16"));
var _uvn5txf = _interopRequireDefault(require("./../.stylex/uvn5txf"));
var _u16kelhn = _interopRequireDefault(require("./../.stylex/u16kelhn"));
var _u90t2fx = _interopRequireDefault(require("./../.stylex/u90t2fx"));
var _u1clkrr = _interopRequireDefault(require("./../.stylex/u1clkrr2"));
var _uyrswl = _interopRequireDefault(require("./../.stylex/uyrswl5"));
var _uwlx52q = _interopRequireDefault(require("./../.stylex/uwlx52q"));
var _u10ismar = _interopRequireDefault(require("./../.stylex/u10ismar"));
var _u1g1nbrb = _interopRequireDefault(require("./../.stylex/u1g1nbrb"));
var _uia0x1d = _interopRequireDefault(require("./../.stylex/uia0x1d"));
var _u135qr3d = _interopRequireDefault(require("./../.stylex/u135qr3d"));
var _u60n6l = _interopRequireDefault(require("./../.stylex/u60n6l"));
var _urjm3ap = _interopRequireDefault(require("./../.stylex/urjm3ap"));
var _u143d86t = _interopRequireDefault(require("./../.stylex/u143d86t"));
var _u1ust6hi = _interopRequireDefault(require("./../.stylex/u1ust6hi"));
var _u1xrgxpq = _interopRequireDefault(require("./../.stylex/u1xrgxpq"));
var _u1773ck = _interopRequireDefault(require("./../.stylex/u1773ck5"));
var _upqf1qz = _interopRequireDefault(require("./../.stylex/upqf1qz"));
var _u1cxaokx = _interopRequireDefault(require("./../.stylex/u1cxaokx"));
var _uzx399i = _interopRequireDefault(require("./../.stylex/uzx399i"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
//.u1rczmw{color:@text-disabled;}
//.u194huds{color:@text-placeholder;}
//.u1z30l8{color:@text-title;}
//.u1lmfglv{margin-top:0px;}
//.uibjyge{margin-right:0px;}
//.u1abmfm4{margin-bottom:0px;}
//.u9yszdx{margin-left:0px;}
//.u82itaf{line-height:22px;}
//.uses1vf{font-size:14px;}
//.undrnj9{font-size:40px;}
//.u150uoa4{padding-bottom:8px;}
//.up6vxd{line-height:1;}
//.u1xowgag{border-top-color:@primary-content-default;}
//.ulspwco{border-right-color:@primary-content-default;}
//.u11qyczi{border-bottom-color:@primary-content-default;}
//.ujldbop{border-left-color:@primary-content-default;}
//.udrzuqr{cursor:not-allowed;}
//.u1k6fduh{cursor:pointer;}
//.u5ogflp{border-top-width:1px;}
//.u1hqa2wf{border-right-width:1px;}
//.u1f09k3d{border-bottom-width:1px;}
//.uinvdd3{border-left-width:1px;}
//.untkm7k{border-top-style:dashed;}
//.u1oq6gla{border-right-style:dashed;}
//.ujbf411{border-bottom-style:dashed;}
//.ulpbmyi{border-left-style:dashed;}
//.ub99si7{border-top-color:@line-border-component;}
//.u12908r0{border-right-color:@line-border-component;}
//.u1lgt5kb{border-bottom-color:@line-border-component;}
//.u3ps6bl{border-left-color:@line-border-component;}
//.uq1byw4{padding-top:32px;}
//.u4v5c5q{padding-bottom:28px;}
//.ujlbh76{padding-left:12px;}
//.u11qrl6u{padding-right:12px;}
//.u17mccla{text-align:center;}
//.u1ipgtue{background-color:@udtoken-component-outlined-bg;}
//.uvo9xlb{border-top-left-radius:@border-radius-md;}
//.ugd3wmz{border-top-right-radius:@border-radius-md;}
//.ueep33z{border-bottom-right-radius:@border-radius-md;}
//.u3qetca{border-bottom-left-radius:@border-radius-md;}
// 拖拽区域
const useDragAreaSx = prefix => {
  return {
    sxDragArea: {
      "props": {
        "Bceei9c": _u12o6wmt.default,
        "B4j52fo": _u1cpcpb.default,
        "Bekrc4i": _u10qn3go.default,
        "Bn0qgzm": _uflkg.default,
        "ibv6hh": _uvn5txf.default,
        "icvyot": _u16kelhn.default,
        "vrafjx": _u90t2fx.default,
        "oivjwe": _u1clkrr.default,
        "wvpqe5": _uyrswl.default,
        "g2u3we": _uwlx52q.default,
        "h3c5rm": _u10ismar.default,
        "B9xav0g": _u1g1nbrb.default,
        "zhjwy3": _uia0x1d.default,
        "z8tnut": _u135qr3d.default,
        "Byoj8tv": _u60n6l.default,
        "uwmqm3": _urjm3ap.default,
        "z189sj": _u143d86t.default,
        "fsow6f": _u1ust6hi.default,
        "De3pzq": _u1xrgxpq.default,
        "Btl43ni": _u1773ck.default,
        "B7oj6ja": _upqf1qz.default,
        "Bbmb7ep": _u1cxaokx.default,
        "Beyfa6y": _uzx399i.default
      },
      "alias": `${prefix}-drag-area`
    },
    sxDisabledText: {
      "props": {
        "sj55zd": _unha2if.default
      }
    },
    sxDragAreaDescription: {
      "props": {
        "sj55zd": _u16q6d1w.default,
        "B6of3ja": _u1kxgv9m.default,
        "t21cq0": _unbze3l.default,
        "jrapky": _u10d06an.default,
        "Frg6f3": _u1w6ojc.default,
        "Bg96gwp": _u1dhzo8t.default,
        "Be2twd7": _u19bueuo.default
      },
      "alias": `${prefix}-drag-area-description`
    },
    sxDragAreaTitle: {
      "props": {
        "sj55zd": _u22yt5q.default,
        "B6of3ja": _u1kxgv9m.default,
        "t21cq0": _unbze3l.default,
        "jrapky": _u10d06an.default,
        "Frg6f3": _u1w6ojc.default,
        "Bg96gwp": _u1dhzo8t.default,
        "Be2twd7": _u19bueuo.default
      },
      "alias": `${prefix}-drag-area-title`
    },
    sxDragAreaIcon: {
      "props": {
        "Be2twd7": _uxmlixj.default,
        "Byoj8tv": _u8ckwbm.default,
        "Bg96gwp": _uv7gtgj.default
      },
      "alias": `${prefix}-drag-area-icon`
    },
    sxDragAreaDisabled: {
      "props": {
        "Bceei9c": _u1vjduc.default
      },
      "alias": `${prefix}-drag-area--disabled`
    },
    sxDragAreaActive: {
      "props": {
        "g2u3we": _u1hkmcjz.default,
        "h3c5rm": _u1hqmxtl.default,
        "B9xav0g": _u1odtry.default,
        "zhjwy3": _ug8e45j.default
      },
      "alias": `${prefix}-drag-area--active`
    }
  };
};
exports.useDragAreaSx = useDragAreaSx;
