// @universe-design/react/lib/upload/UploadTrigger.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UploadTrigger = void 0;
var _u8wgnjj = _interopRequireDefault(require("./../.stylex/u8wgnjj"));
var _u1c70cyo = _interopRequireDefault(require("./../.stylex/u1c70cyo"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = require("@universe-design/stylex/react");
var React = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _uuid = require("uuid");
var _UploadContext = require("./hooks/UploadContext");
var _traverseFileTree = _interopRequireDefault(require("./utils/traverse-file-tree"));
var _forwardRef = require("../util/forwardRef");
var _attrAccept = _interopRequireDefault(require("./utils/attr-accept"));
var _noop = _interopRequireDefault(require("lodash/noop"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
//.u14t3ns0{display:inline-block;}
//.u1s6fcnf{outline-style:none;}
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
/**
 * @public
 * @docgen
 */
/**
 * 触发文件选择
 * 发送请求都在这个组件完成
 * @public
 */
const UploadTrigger = exports.UploadTrigger = (0, _forwardRef.forwardHTMLRef)((props, ref) => {
  const {
    children,
    className,
    onClick,
    onKeyDown,
    openFileDialogOnClick = true,
    type = 'click',
    accept = '',
    directory,
    multiple = false,
    ...htmlProps
  } = props;
  const {
    rootPrefixCls,
    disabled,
    uploadFiles
  } = (0, _UploadContext.useUploadContext)();
  const inputRef = React.useRef(null);
  const [uid, setUid] = React.useState((0, _uuid.v4)());
  const onChange = React.useCallback(e => {
    const {
      files
    } = e.target;
    const acceptedFiles = [...(files !== null && files !== void 0 ? files : [])].filter(file => !directory || (0, _attrAccept.default)(file, accept || ''));
    uploadFiles(acceptedFiles);
    // reset
    setUid((0, _uuid.v4)());
  }, [accept, directory, uploadFiles]);

  // because input don't have directory/webkitdirectory type declaration
  const dirProps = directory ? {
    directory: 'directory',
    webkitdirectory: 'webkitdirectory'
  } : {};
  const handleClick = React.useCallback(e => {
    var _inputRef$current;
    (_inputRef$current = inputRef.current) === null || _inputRef$current === void 0 ? void 0 : _inputRef$current.click();
    onClick === null || onClick === void 0 ? void 0 : onClick(e);
  }, [onClick]);
  const handleKeyDown = React.useCallback(e => {
    if (e.key === 'Enter') {
      var _inputRef$current2;
      (_inputRef$current2 = inputRef.current) === null || _inputRef$current2 === void 0 ? void 0 : _inputRef$current2.click();
    }
    onKeyDown === null || onKeyDown === void 0 ? void 0 : onKeyDown(e);
  }, [onKeyDown]);
  const handleDrop = React.useCallback(e => {
    e.preventDefault();
    if (e.type === 'dragover') {
      return;
    }
    if (directory) {
      (0, _traverseFileTree.default)(Array.prototype.slice.call(e.dataTransfer.items), uploadFiles, _file => (0, _attrAccept.default)(_file, accept));
    } else {
      const files = [...e.dataTransfer.files].filter(file => (0, _attrAccept.default)(file, accept));
      if (!multiple) {
        files.splice(1);
      }
      uploadFiles(files);
    }
  }, [directory, uploadFiles, accept, multiple]);
  const events = React.useMemo(() => {
    if (disabled) {
      return {};
    }
    return {
      onClick: openFileDialogOnClick ? handleClick : _noop.default,
      onKeyDown: openFileDialogOnClick ? handleKeyDown : _noop.default,
      onDrop: type === 'drag' ? handleDrop : _noop.default,
      onDragOver: type === 'drag' ? handleDrop : _noop.default,
      tabIndex: 0
    };
  }, [disabled, handleClick, handleDrop, handleKeyDown, openFileDialogOnClick, type]);
  return (0, _react.jsx)("div", (0, _extends2.default)({
    className: (0, _clsx.default)(`${rootPrefixCls}-trigger`, className),
    ref: ref
  }, events, htmlProps, {
    sx: [{
      "props": {
        "oeaueh": _u1c70cyo.default
      }
    }, type === 'click' && {
      "props": {
        "mc9l5x": _u8wgnjj.default
      }
    }]
  }), (0, _react.jsx)("input", (0, _extends2.default)({
    type: "file",
    ref: inputRef,
    onClick: e => e.stopPropagation(),
    key: uid,
    style: {
      display: 'none'
    },
    accept: accept
  }, dirProps, {
    multiple: multiple,
    onChange: onChange
  })), children);
});
