// @universe-design/react/lib/upload/Upload.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Upload = void 0;
var _uiqudwx = _interopRequireDefault(require("./../.stylex/uiqudwx"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _ComposableUpload = require("./ComposableUpload");
var _uploadFile = require("./uploadFile");
var _uploadList = require("./uploadList");
var _PictureUploadTrigger = require("./uploadList/PictureUploadTrigger");
var _UploadTrigger = require("./UploadTrigger");
var _utils = require("./utils");
var _UploadContext = require("./hooks/UploadContext");
var _useUploadFile = require("./uploadFile/useUploadFile");
//.u2zxqi{margin-top:8px;}
const FileItem = _ref => {
  let {
    itemRender,
    listType,
    fileProps
  } = _ref;
  const {
    file
  } = fileProps;
  const fileList = (0, _UploadContext.useUploadContextSelector)(ctx => ctx.fileList);
  let OriginComp = _uploadFile.UploadTextFile;
  switch (listType) {
    case 'card':
      {
        OriginComp = _uploadFile.UploadCardFile;
        break;
      }
    case 'picture':
      {
        OriginComp = _uploadFile.UploadPictureFile;
        break;
      }
    case 'simple-text':
      {
        OriginComp = _uploadFile.UploadSimpleTextFile;
        break;
      }
    default:
      {
        OriginComp = _uploadFile.UploadTextFile;
        break;
      }
  }
  const fileCtx = (0, _useUploadFile.useUploadFile)(fileProps);
  const actions = {
    download: fileCtx.onDownload,
    preview: fileCtx.onPreview,
    remove: fileCtx.onRemove,
    retry: fileCtx.onRetry
  };
  const originNode = /*#__PURE__*/React.createElement(OriginComp, fileProps);
  if (itemRender) {
    return /*#__PURE__*/React.createElement(React.Fragment, null, itemRender(originNode, file, fileList, actions));
  }
  return originNode;
};

/**
 *  推荐使用 `ComposableUpload` 来实现更灵活的组合式用法
 */
const Upload = exports.Upload = /*#__PURE__*/React.forwardRef((props, ref) => {
  const {
    defaultFileList,
    fileList,
    disabled,
    onChange,
    maxCount,
    onExceed,
    children,
    listType = 'text',
    tips,
    showUploadList = true,
    showPictureUploadButton = true,
    className,
    style,
    // triggerProps start
    method,
    multiple,
    accept = listType === 'picture' ? 'image/*' : '',
    action,
    beforeUpload,
    customRequest,
    data,
    directory,
    headers,
    name,
    openFileDialogOnClick,
    type,
    withCredentials,
    // uploadList
    itemRender: itemRenderProp,
    maxDisplayedPictureCount,
    onPreview,
    onRetry,
    onRemove,
    onDownload,
    // triggerProps end
    ...uploadListProps
  } = props;
  const isPictureUpload = listType === 'picture';
  const uploadListIcons = typeof showUploadList !== 'object' ? {} : showUploadList;
  const uploadTriggerProps = {
    openFileDialogOnClick,
    type,
    accept,
    multiple,
    directory
  };
  const getFileProps = file => ({
    ...uploadListProps,
    ...uploadListIcons,
    file,
    ...(onPreview ? {
      onPreview: () => onPreview(file)
    } : {}),
    ...(onRetry ? {
      onRetry: () => onRetry(file)
    } : {}),
    ...(onRemove ? {
      onRemove: () => onRemove(file)
    } : {}),
    ...(onDownload ? {
      onDownload: () => onDownload(file)
    } : {})
  });
  return /*#__PURE__*/React.createElement(_ComposableUpload.ComposableUpload, (0, _extends2.default)({}, {
    defaultFileList,
    fileList,
    disabled,
    onChange,
    maxCount,
    onExceed,
    className,
    style,
    method,
    action,
    beforeUpload,
    customRequest,
    data,
    headers,
    name,
    openFileDialogOnClick,
    type,
    withCredentials
  }, {
    ref: ref
  }), !isPictureUpload && /*#__PURE__*/React.createElement(_UploadTrigger.UploadTrigger, uploadTriggerProps, children), !isPictureUpload && tips, showUploadList && (() => {
    const classNames = children ? (0, _stylex.default)({
      "props": {
        "B6of3ja": _uiqudwx.default
      }
    }) : undefined;
    const itemRenderer = file => /*#__PURE__*/React.createElement(FileItem, {
      listType: listType,
      fileProps: getFileProps(file),
      itemRender: itemRenderProp
    });
    switch (listType) {
      case 'card':
        return /*#__PURE__*/React.createElement(_uploadList.UploadCardFileList, {
          className: classNames,
          itemRenderer: itemRenderer
        });
      case 'picture':
        return /*#__PURE__*/React.createElement(_uploadList.UploadPictureFileList, {
          maxDisplayedPictureCount: maxDisplayedPictureCount,
          uploadTrigger: showPictureUploadButton && /*#__PURE__*/React.createElement(_PictureUploadTrigger.PictureUploadTrigger, uploadTriggerProps),
          itemRenderer: itemRenderer
        });
      case 'simple-text':
        return /*#__PURE__*/React.createElement(_uploadList.UploadSimpleTextFileList, {
          className: classNames,
          itemRenderer: itemRenderer
        });
      default:
        return /*#__PURE__*/React.createElement(_uploadList.UploadTextFileList, {
          className: classNames,
          itemRenderer: itemRenderer
        });
    }
  })(), isPictureUpload && tips);
});
Upload.LIST_IGNORE = _utils.LIST_IGNORE;
if (process.env.NODE_ENV !== "production") {
  Upload.displayName = 'Upload';
}
