// @universe-design/react/es/upload/Dragger.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React, { useState } from 'react';
import clsx from 'clsx';
import stylex from '@universe-design/stylex';
import FileAddColorful from '@universe-design/icons-react/FileAddColorful';
import FileAddDisableColorful from '@universe-design/icons-react/FileAddDisableColorful';
import { Upload } from './Upload';
import { useUploadContext } from './hooks/UploadContext';
import { useDragAreaSx } from './sx';
import { forwardHTMLRef } from '../util/forwardRef';
export const UploadDragArea = forwardHTMLRef((props, ref) => {
  const {
    dragIcon: dragIconProp,
    dragDescription,
    dragTitle,
    className,
    onMouseEnter: onMouseEnterProp,
    onMouseLeave: onMouseLeaveProp,
    onDragLeave: onDragLeaveProp,
    onDragOver: onDragOverProp,
    ...resetProps
  } = props;
  const {
    rootPrefixCls,
    disabled
  } = useUploadContext();
  const [isActive, setIsActive] = useState(false);
  const {
    sxDisabledText,
    sxDragArea,
    sxDragAreaActive,
    sxDragAreaDescription,
    sxDragAreaDisabled,
    sxDragAreaIcon,
    sxDragAreaTitle
  } = useDragAreaSx(rootPrefixCls);
  const classNames = clsx([className, stylex(sxDragArea, isActive && sxDragAreaActive, disabled && sxDragAreaDisabled)]);
  const dragIcon = dragIconProp !== null && dragIconProp !== void 0 ? dragIconProp : disabledArg => disabledArg ? /*#__PURE__*/React.createElement(FileAddDisableColorful, null) : /*#__PURE__*/React.createElement(FileAddColorful, null);
  const icon = typeof dragIcon === 'function' ? dragIcon(disabled) : dragIcon;
  const events = React.useMemo(() => {
    if (disabled) {
      return {};
    }
    const active = () => setIsActive(true);
    const inactive = () => setIsActive(false);
    return {
      onMouseEnter: e => {
        onMouseEnterProp === null || onMouseEnterProp === void 0 ? void 0 : onMouseEnterProp(e);
        active();
      },
      onMouseLeave: e => {
        onMouseLeaveProp === null || onMouseLeaveProp === void 0 ? void 0 : onMouseLeaveProp(e);
        inactive();
      },
      onDragLeave: e => {
        onDragLeaveProp === null || onDragLeaveProp === void 0 ? void 0 : onDragLeaveProp(e);
        inactive();
      },
      onDragOver: e => {
        onDragOverProp === null || onDragOverProp === void 0 ? void 0 : onDragOverProp(e);
        active();
      }
    };
  }, [disabled, onDragLeaveProp, onDragOverProp, onMouseEnterProp, onMouseLeaveProp]);
  return /*#__PURE__*/React.createElement("div", _extends({
    className: classNames,
    ref: ref
  }, resetProps, events), /*#__PURE__*/React.createElement("div", {
    className: stylex(sxDragAreaIcon)
  }, icon), /*#__PURE__*/React.createElement("div", {
    className: stylex(sxDragAreaTitle, disabled && sxDisabledText)
  }, dragTitle), /*#__PURE__*/React.createElement("p", {
    className: clsx(stylex(sxDragAreaDescription, disabled && sxDisabledText))
  }, dragDescription));
});
const InternalDragger = (_ref, ref) => {
  let {
    dragAreaConfig,
    ...restProps
  } = _ref;
  return /*#__PURE__*/React.createElement(Upload, _extends({
    ref: ref
  }, restProps, {
    type: "drag"
  }), dragAreaConfig ? /*#__PURE__*/React.createElement(UploadDragArea, dragAreaConfig) : restProps.children);
};
const Dragger = /*#__PURE__*/React.forwardRef(InternalDragger);
if (process.env.NODE_ENV !== "production") {
  Dragger.displayName = 'Dragger';
}
export default Dragger;
