// @universe-design/react/es/upload/sx.js
//.u1rczmw{color:@text-disabled;}
import unha2if from "./../.stylex/unha2if";
//.u194huds{color:@text-placeholder;}
import u16q6d1w from "./../.stylex/u16q6d1w";
//.u1z30l8{color:@text-title;}
import u22yt5q from "./../.stylex/u22yt5q";
//.u1lmfglv{margin-top:0px;}
import u1kxgv9m from "./../.stylex/u1kxgv9m";
//.uibjyge{margin-right:0px;}
import unbze3l from "./../.stylex/unbze3l";
//.u1abmfm4{margin-bottom:0px;}
import u10d06an from "./../.stylex/u10d06an";
//.u9yszdx{margin-left:0px;}
import u1w6ojc6 from "./../.stylex/u1w6ojc6";
//.u82itaf{line-height:22px;}
import u1dhzo8t from "./../.stylex/u1dhzo8t";
//.uses1vf{font-size:14px;}
import u19bueuo from "./../.stylex/u19bueuo";
//.undrnj9{font-size:40px;}
import uxmlixj from "./../.stylex/uxmlixj";
//.u150uoa4{padding-bottom:8px;}
import u8ckwbm from "./../.stylex/u8ckwbm";
//.up6vxd{line-height:1;}
import uv7gtgj from "./../.stylex/uv7gtgj";
//.u1xowgag{border-top-color:@primary-content-default;}
import u1hkmcjz from "./../.stylex/u1hkmcjz";
//.ulspwco{border-right-color:@primary-content-default;}
import u1hqmxtl from "./../.stylex/u1hqmxtl";
//.u11qyczi{border-bottom-color:@primary-content-default;}
import u1odtry2 from "./../.stylex/u1odtry2";
//.ujldbop{border-left-color:@primary-content-default;}
import ug8e45j from "./../.stylex/ug8e45j";
//.udrzuqr{cursor:not-allowed;}
import u1vjduc2 from "./../.stylex/u1vjduc2";
//.u1k6fduh{cursor:pointer;}
import u12o6wmt from "./../.stylex/u12o6wmt";
//.u5ogflp{border-top-width:1px;}
import u1cpcpb5 from "./../.stylex/u1cpcpb5";
//.u1hqa2wf{border-right-width:1px;}
import u10qn3go from "./../.stylex/u10qn3go";
//.u1f09k3d{border-bottom-width:1px;}
import uflkg16 from "./../.stylex/uflkg16";
//.uinvdd3{border-left-width:1px;}
import uvn5txf from "./../.stylex/uvn5txf";
//.untkm7k{border-top-style:dashed;}
import u16kelhn from "./../.stylex/u16kelhn";
//.u1oq6gla{border-right-style:dashed;}
import u90t2fx from "./../.stylex/u90t2fx";
//.ujbf411{border-bottom-style:dashed;}
import u1clkrr2 from "./../.stylex/u1clkrr2";
//.ulpbmyi{border-left-style:dashed;}
import uyrswl5 from "./../.stylex/uyrswl5";
//.ub99si7{border-top-color:@line-border-component;}
import uwlx52q from "./../.stylex/uwlx52q";
//.u12908r0{border-right-color:@line-border-component;}
import u10ismar from "./../.stylex/u10ismar";
//.u1lgt5kb{border-bottom-color:@line-border-component;}
import u1g1nbrb from "./../.stylex/u1g1nbrb";
//.u3ps6bl{border-left-color:@line-border-component;}
import uia0x1d from "./../.stylex/uia0x1d";
//.uq1byw4{padding-top:32px;}
import u135qr3d from "./../.stylex/u135qr3d";
//.u4v5c5q{padding-bottom:28px;}
import u60n6l from "./../.stylex/u60n6l";
//.ujlbh76{padding-left:12px;}
import urjm3ap from "./../.stylex/urjm3ap";
//.u11qrl6u{padding-right:12px;}
import u143d86t from "./../.stylex/u143d86t";
//.u17mccla{text-align:center;}
import u1ust6hi from "./../.stylex/u1ust6hi";
//.u1ipgtue{background-color:@udtoken-component-outlined-bg;}
import u1xrgxpq from "./../.stylex/u1xrgxpq";
//.uvo9xlb{border-top-left-radius:@border-radius-md;}
import u1773ck5 from "./../.stylex/u1773ck5";
//.ugd3wmz{border-top-right-radius:@border-radius-md;}
import upqf1qz from "./../.stylex/upqf1qz";
//.ueep33z{border-bottom-right-radius:@border-radius-md;}
import u1cxaokx from "./../.stylex/u1cxaokx";
//.u3qetca{border-bottom-left-radius:@border-radius-md;}
import uzx399i from "./../.stylex/uzx399i";
import stylex from '@universe-design/stylex';
// 拖拽区域
export const useDragAreaSx = prefix => {
  return {
    sxDragArea: {
      "props": {
        "Bceei9c": u12o6wmt,
        "B4j52fo": u1cpcpb5,
        "Bekrc4i": u10qn3go,
        "Bn0qgzm": uflkg16,
        "ibv6hh": uvn5txf,
        "icvyot": u16kelhn,
        "vrafjx": u90t2fx,
        "oivjwe": u1clkrr2,
        "wvpqe5": uyrswl5,
        "g2u3we": uwlx52q,
        "h3c5rm": u10ismar,
        "B9xav0g": u1g1nbrb,
        "zhjwy3": uia0x1d,
        "z8tnut": u135qr3d,
        "Byoj8tv": u60n6l,
        "uwmqm3": urjm3ap,
        "z189sj": u143d86t,
        "fsow6f": u1ust6hi,
        "De3pzq": u1xrgxpq,
        "Btl43ni": u1773ck5,
        "B7oj6ja": upqf1qz,
        "Bbmb7ep": u1cxaokx,
        "Beyfa6y": uzx399i
      },
      "alias": `${prefix}-drag-area`
    },
    sxDisabledText: {
      "props": {
        "sj55zd": unha2if
      }
    },
    sxDragAreaDescription: {
      "props": {
        "sj55zd": u16q6d1w,
        "B6of3ja": u1kxgv9m,
        "t21cq0": unbze3l,
        "jrapky": u10d06an,
        "Frg6f3": u1w6ojc6,
        "Bg96gwp": u1dhzo8t,
        "Be2twd7": u19bueuo
      },
      "alias": `${prefix}-drag-area-description`
    },
    sxDragAreaTitle: {
      "props": {
        "sj55zd": u22yt5q,
        "B6of3ja": u1kxgv9m,
        "t21cq0": unbze3l,
        "jrapky": u10d06an,
        "Frg6f3": u1w6ojc6,
        "Bg96gwp": u1dhzo8t,
        "Be2twd7": u19bueuo
      },
      "alias": `${prefix}-drag-area-title`
    },
    sxDragAreaIcon: {
      "props": {
        "Be2twd7": uxmlixj,
        "Byoj8tv": u8ckwbm,
        "Bg96gwp": uv7gtgj
      },
      "alias": `${prefix}-drag-area-icon`
    },
    sxDragAreaDisabled: {
      "props": {
        "Bceei9c": u1vjduc2
      },
      "alias": `${prefix}-drag-area--disabled`
    },
    sxDragAreaActive: {
      "props": {
        "g2u3we": u1hkmcjz,
        "h3c5rm": u1hqmxtl,
        "B9xav0g": u1odtry2,
        "zhjwy3": ug8e45j
      },
      "alias": `${prefix}-drag-area--active`
    }
  };
};
