// @universe-design/react/es/upload/UploadTrigger.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.u14t3ns0{display:inline-block;}
import u8wgnjj from "./../.stylex/u8wgnjj";
//.u1s6fcnf{outline-style:none;}
import u1c70cyo from "./../.stylex/u1c70cyo";
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx } from '@universe-design/stylex/react';
import * as React from 'react';
import clsx from 'clsx';
import { v4 as uuid } from 'uuid';
import { useUploadContext } from './hooks/UploadContext';
import traverseFileTree from './utils/traverse-file-tree';
import { forwardHTMLRef } from '../util/forwardRef';
import attrAccept from './utils/attr-accept';
import noop from 'lodash/noop';
import stylex from '@universe-design/stylex';

/**
 * @public
 * @docgen
 */

/**
 * 触发文件选择
 * 发送请求都在这个组件完成
 * @public
 */
export const UploadTrigger = forwardHTMLRef((props, ref) => {
  const {
    children,
    className,
    onClick,
    onKeyDown,
    openFileDialogOnClick = true,
    type = 'click',
    accept = '',
    directory,
    multiple = false,
    ...htmlProps
  } = props;
  const {
    rootPrefixCls,
    disabled,
    uploadFiles
  } = useUploadContext();
  const inputRef = React.useRef(null);
  const [uid, setUid] = React.useState(uuid());
  const onChange = React.useCallback(e => {
    const {
      files
    } = e.target;
    const acceptedFiles = [...(files !== null && files !== void 0 ? files : [])].filter(file => !directory || attrAccept(file, accept || ''));
    uploadFiles(acceptedFiles);
    // reset
    setUid(uuid());
  }, [accept, directory, uploadFiles]);

  // because input don't have directory/webkitdirectory type declaration
  const dirProps = directory ? {
    directory: 'directory',
    webkitdirectory: 'webkitdirectory'
  } : {};
  const handleClick = React.useCallback(e => {
    var _inputRef$current;
    (_inputRef$current = inputRef.current) === null || _inputRef$current === void 0 ? void 0 : _inputRef$current.click();
    onClick === null || onClick === void 0 ? void 0 : onClick(e);
  }, [onClick]);
  const handleKeyDown = React.useCallback(e => {
    if (e.key === 'Enter') {
      var _inputRef$current2;
      (_inputRef$current2 = inputRef.current) === null || _inputRef$current2 === void 0 ? void 0 : _inputRef$current2.click();
    }
    onKeyDown === null || onKeyDown === void 0 ? void 0 : onKeyDown(e);
  }, [onKeyDown]);
  const handleDrop = React.useCallback(e => {
    e.preventDefault();
    if (e.type === 'dragover') {
      return;
    }
    if (directory) {
      traverseFileTree(Array.prototype.slice.call(e.dataTransfer.items), uploadFiles, _file => attrAccept(_file, accept));
    } else {
      const files = [...e.dataTransfer.files].filter(file => attrAccept(file, accept));
      if (!multiple) {
        files.splice(1);
      }
      uploadFiles(files);
    }
  }, [directory, uploadFiles, accept, multiple]);
  const events = React.useMemo(() => {
    if (disabled) {
      return {};
    }
    return {
      onClick: openFileDialogOnClick ? handleClick : noop,
      onKeyDown: openFileDialogOnClick ? handleKeyDown : noop,
      onDrop: type === 'drag' ? handleDrop : noop,
      onDragOver: type === 'drag' ? handleDrop : noop,
      tabIndex: 0
    };
  }, [disabled, handleClick, handleDrop, handleKeyDown, openFileDialogOnClick, type]);
  return jsx("div", _extends({
    className: clsx(`${rootPrefixCls}-trigger`, className),
    ref: ref
  }, events, htmlProps, {
    sx: [{
      "props": {
        "oeaueh": u1c70cyo
      }
    }, type === 'click' && {
      "props": {
        "mc9l5x": u8wgnjj
      }
    }]
  }), jsx("input", _extends({
    type: "file",
    ref: inputRef,
    onClick: e => e.stopPropagation(),
    key: uid,
    style: {
      display: 'none'
    },
    accept: accept
  }, dirProps, {
    multiple: multiple,
    onChange: onChange
  })), children);
});
