// @universe-design/react/es/upload/Upload.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.u2zxqi{margin-top:8px;}
import uiqudwx from "./../.stylex/uiqudwx";
import * as React from 'react';
import stylex from '@universe-design/stylex';
import { ComposableUpload } from './ComposableUpload';
import { UploadCardFile, UploadPictureFile, UploadSimpleTextFile, UploadTextFile } from './uploadFile';
import { UploadCardFileList, UploadPictureFileList, UploadSimpleTextFileList, UploadTextFileList } from './uploadList';
import { PictureUploadTrigger } from './uploadList/PictureUploadTrigger';
import { UploadTrigger } from './UploadTrigger';
import { LIST_IGNORE } from './utils';
import { useUploadContextSelector } from './hooks/UploadContext';
import { useUploadFile } from './uploadFile/useUploadFile';
const FileItem = _ref => {
  let {
    itemRender,
    listType,
    fileProps
  } = _ref;
  const {
    file
  } = fileProps;
  const fileList = useUploadContextSelector(ctx => ctx.fileList);
  let OriginComp = UploadTextFile;
  switch (listType) {
    case 'card':
      {
        OriginComp = UploadCardFile;
        break;
      }
    case 'picture':
      {
        OriginComp = UploadPictureFile;
        break;
      }
    case 'simple-text':
      {
        OriginComp = UploadSimpleTextFile;
        break;
      }
    default:
      {
        OriginComp = UploadTextFile;
        break;
      }
  }
  const fileCtx = useUploadFile(fileProps);
  const actions = {
    download: fileCtx.onDownload,
    preview: fileCtx.onPreview,
    remove: fileCtx.onRemove,
    retry: fileCtx.onRetry
  };
  const originNode = /*#__PURE__*/React.createElement(OriginComp, fileProps);
  if (itemRender) {
    return /*#__PURE__*/React.createElement(React.Fragment, null, itemRender(originNode, file, fileList, actions));
  }
  return originNode;
};

/**
 *  推荐使用 `ComposableUpload` 来实现更灵活的组合式用法
 */
export const Upload = /*#__PURE__*/React.forwardRef((props, ref) => {
  const {
    defaultFileList,
    fileList,
    disabled,
    onChange,
    maxCount,
    onExceed,
    children,
    listType = 'text',
    tips,
    showUploadList = true,
    showPictureUploadButton = true,
    className,
    style,
    // triggerProps start
    method,
    multiple,
    accept = listType === 'picture' ? 'image/*' : '',
    action,
    beforeUpload,
    customRequest,
    data,
    directory,
    headers,
    name,
    openFileDialogOnClick,
    type,
    withCredentials,
    // uploadList
    itemRender: itemRenderProp,
    maxDisplayedPictureCount,
    onPreview,
    onRetry,
    onRemove,
    onDownload,
    // triggerProps end
    ...uploadListProps
  } = props;
  const isPictureUpload = listType === 'picture';
  const uploadListIcons = typeof showUploadList !== 'object' ? {} : showUploadList;
  const uploadTriggerProps = {
    openFileDialogOnClick,
    type,
    accept,
    multiple,
    directory
  };
  const getFileProps = file => ({
    ...uploadListProps,
    ...uploadListIcons,
    file,
    ...(onPreview ? {
      onPreview: () => onPreview(file)
    } : {}),
    ...(onRetry ? {
      onRetry: () => onRetry(file)
    } : {}),
    ...(onRemove ? {
      onRemove: () => onRemove(file)
    } : {}),
    ...(onDownload ? {
      onDownload: () => onDownload(file)
    } : {})
  });
  return /*#__PURE__*/React.createElement(ComposableUpload, _extends({}, {
    defaultFileList,
    fileList,
    disabled,
    onChange,
    maxCount,
    onExceed,
    className,
    style,
    method,
    action,
    beforeUpload,
    customRequest,
    data,
    headers,
    name,
    openFileDialogOnClick,
    type,
    withCredentials
  }, {
    ref: ref
  }), !isPictureUpload && /*#__PURE__*/React.createElement(UploadTrigger, uploadTriggerProps, children), !isPictureUpload && tips, showUploadList && (() => {
    const classNames = children ? stylex({
      "props": {
        "B6of3ja": uiqudwx
      }
    }) : undefined;
    const itemRenderer = file => /*#__PURE__*/React.createElement(FileItem, {
      listType: listType,
      fileProps: getFileProps(file),
      itemRender: itemRenderProp
    });
    switch (listType) {
      case 'card':
        return /*#__PURE__*/React.createElement(UploadCardFileList, {
          className: classNames,
          itemRenderer: itemRenderer
        });
      case 'picture':
        return /*#__PURE__*/React.createElement(UploadPictureFileList, {
          maxDisplayedPictureCount: maxDisplayedPictureCount,
          uploadTrigger: showPictureUploadButton && /*#__PURE__*/React.createElement(PictureUploadTrigger, uploadTriggerProps),
          itemRenderer: itemRenderer
        });
      case 'simple-text':
        return /*#__PURE__*/React.createElement(UploadSimpleTextFileList, {
          className: classNames,
          itemRenderer: itemRenderer
        });
      default:
        return /*#__PURE__*/React.createElement(UploadTextFileList, {
          className: classNames,
          itemRenderer: itemRenderer
        });
    }
  })(), isPictureUpload && tips);
});
Upload.LIST_IGNORE = LIST_IGNORE;
if (process.env.NODE_ENV !== "production") {
  Upload.displayName = 'Upload';
}
