// @universe-design/react/es/upload/utils/index.js
import isNil from 'lodash/isNil';
import clone from 'lodash/clone';
export const LIST_IGNORE = `__LIST_IGNORE_${Date.now()}__`;
export function formatFileSize(bytes, locale) {
  if (isNil(bytes)) {
    return locale.unknownSize;
  }
  if (bytes < 1024) {
    return `${bytes}B`;
  }
  const kbs = bytes / 1024;
  if (kbs < 1024) {
    return `${kbs.toFixed(2)}KB`;
  }
  const ms = kbs / 1024;
  if (ms < 1024) {
    return `${ms.toFixed(2)}MB`;
  }
  return `${(ms / 1024).toFixed(2)}GB`;
}
export function getFileItem(file, fileList) {
  const matchKey = file.uid !== undefined ? 'uid' : 'name';
  return fileList.filter(item => item[matchKey] === file[matchKey])[0];
}
export const file2Obj = file => ({
  ...file,
  lastModified: file.lastModified,
  lastModifiedDate: file.lastModifiedDate,
  name: file.name,
  size: file.size,
  type: file.type,
  uid: file.uid,
  percent: 0,
  originFileObj: file
});
export const updateFileList = (file, fileList) => {
  const nextFileList = [...fileList];
  const clonedFile = clone(file);
  const fileIndex = nextFileList.findIndex(_ref => {
    let {
      uid
    } = _ref;
    return uid === clonedFile.uid;
  });
  if (fileIndex === -1) {
    nextFileList.push(clonedFile);
  } else {
    nextFileList[fileIndex] = clonedFile;
  }
  return nextFileList;
};
export function removeFileItem(file, fileList) {
  const matchKey = file.uid !== undefined ? 'uid' : 'name';
  const removed = fileList.filter(item => item[matchKey] !== file[matchKey]);
  if (removed.length === fileList.length) {
    return null;
  }
  return removed;
}
const isImageFileType = type => type.indexOf('image/') === 0;
const extname = function () {
  let url = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
  const temp = url.split('/');
  const filename = temp[temp.length - 1];
  const filenameWithoutSuffix = filename.split(/#|\?/)[0];
  return (/\.[^./\\]*$/.exec(filenameWithoutSuffix) || [''])[0];
};
export function isImage(file) {
  if (file.thumbUrl) {
    return true;
  }
  if (file.type) {
    return isImageFileType(file.type);
  }
  const extension = extname(file.url);
  if (/^data:image\//.test(file.url || '') || /(webp|svg|png|gif|jpg|jpeg|jfif|bmp|dpg|ico)$/i.test(extension)) {
    return true;
  }
  return false;
}
export function previewImage(file) {
  const MEASURE_SIZE = 200;
  return new Promise(resolve => {
    if (!file.type || !isImageFileType(file.type)) {
      resolve('');
      return;
    }
    const canvas = document.createElement('canvas');
    canvas.width = MEASURE_SIZE;
    canvas.height = MEASURE_SIZE;
    canvas.style.cssText = `position: fixed; left: 0; top: 0; width: ${MEASURE_SIZE}px;` + `height: ${MEASURE_SIZE}px; z-index: 9999; display: none;`;
    document.body.appendChild(canvas);
    const ctx = canvas.getContext('2d');
    const img = new Image();
    img.onload = () => {
      const {
        width,
        height
      } = img;
      let drawWidth = MEASURE_SIZE;
      let drawHeight = MEASURE_SIZE;
      let offsetX = 0;
      let offsetY = 0;
      if (width > height) {
        drawHeight = height * (MEASURE_SIZE / width);
        offsetY = -(drawHeight - drawWidth) / 2;
      } else {
        drawWidth = width * (MEASURE_SIZE / height);
        offsetX = -(drawWidth - drawHeight) / 2;
      }
      ctx.drawImage(img, offsetX, offsetY, drawWidth, drawHeight);
      const dataURL = canvas.toDataURL();
      document.body.removeChild(canvas);
      resolve(dataURL);
    };
    img.src = window.URL.createObjectURL(file);
  });
}
export function download(url, filename) {
  const element = document.createElement('a');
  element.setAttribute('href', url);
  element.setAttribute('target', '_blank');
  element.setAttribute('rel', 'noopener noreferrer');
  element.setAttribute('download', filename);
  document.body.appendChild(element);
  element.click();
  document.body.removeChild(element);
}
export function internalDownload(file) {
  if (file.url) {
    download(file.url, file.name);
  }
}
export function internalPreview(file) {
  const url = file.url || file.thumbUrl;
  if (url) {
    window.open(url);
  }
}

/**
 * 是否处于 error 态
 */
export function isError(file) {
  return file.status === 'downloadFailed' || file.status === 'error';
}
