// @byted-nodex/metrics/lib/collector/collector.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Collector = void 0;
const consts_1 = require("../consts");
const metrics_codec_1 = require("@byted-nodex/metrics-codec");
const env_1 = __importDefault(require("@byted-service/env"));
const util_1 = require("../util");
class Collector {
    constructor(metric) {
        this.metric = metric;
        this.size = 0;
    }
    alloc(sendTime) {
        const message = new metrics_codec_1.Message({
            tenant: this.metric.tenant,
            project: env_1.default.getPSM(),
            sdkVersion: (0, util_1.getSdkVersion)(),
        });
        message.setTimestamp(sendTime);
        message.setFlags(0x07);
        return message;
    }
    /** 卸货 */
    async flush(sendTime, send) {
        const summary = {
            packedSize: 0,
            cacheCursorCount: this.size,
            packedCursorCount: 0,
            cacheMetricCount: 0,
            packedMetricCount: 0,
        };
        await this.pack(sendTime, summary, async (m) => {
            const packet = m.encode();
            summary.packedSize += packet.byteLength;
            // 打包完立马 reset 掉节省打包时的资源占用
            m.reset();
            await send(packet);
        });
        return summary;
    }
    /** 打包 */
    async pack(sendTime, summary, cb) {
        const sendTimeSec = Math.round(sendTime / 1000);
        const timestamp = BigInt(sendTimeSec);
        const lastTimeStamp = BigInt(sendTimeSec - this.metric.valuePolicy.interval);
        let message = this.alloc(timestamp);
        let lastMessage = this.alloc(lastTimeStamp);
        const cursorList = [];
        // collect cursors
        this.each(cursor => {
            if (cursor.active) {
                cursorList.push(cursor);
            }
            summary.cacheMetricCount += cursor.values.size;
        });
        for (const cursor of cursorList) {
            const result = cursor.pack(message, lastMessage);
            if (!result) {
                continue;
            }
            summary.packedCursorCount++;
            summary.packedMetricCount += result;
            if (message.byteLength >= consts_1.oneMessageSizeLimit) {
                await cb(message);
                message = this.alloc(timestamp);
            }
            if (lastMessage.byteLength >= consts_1.oneMessageSizeLimit) {
                await cb(lastMessage);
                lastMessage = this.alloc(lastTimeStamp);
            }
        }
        if (message.metrics.length) {
            await cb(message);
        }
        if (lastMessage.metrics.length) {
            await cb(lastMessage);
        }
    }
    analyse(info) {
        info.cursorCount += this.size;
        this.each(cursor => {
            cursor.values.analyse(info);
        });
    }
}
exports.Collector = Collector;
