// @byted-nodex/metrics/lib/collector/kdtree.js
"use strict";
/**
 * 通过 kdtree + 分片来索引序列，比较节省内存也比较快
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.KdTreeCollector = exports.KdTree = void 0;
const collector_1 = require("./collector");
const cursor_1 = require("./cursor");
const consts_1 = require("../consts");
const util_1 = require("../util");
function newNode(c, left, right) {
    return {
        value: c,
        left,
        right,
    };
}
function isMatch(n, index) {
    for (let i = 0; i < n.value.index.length; i++) {
        if (index[i] !== n.value.index[i])
            return false;
    }
    return true;
}
class KdTree {
    constructor(metric, counter) {
        this.d = metric.sortedTags.length;
        this.counter = counter || { size: 0, maxCursorSize: consts_1.defaultMaxCursorSize };
        this.metric = metric;
    }
    createNode(index) {
        if (this.counter.size >= this.counter.maxCursorSize) {
            this.metric.logger.error('Reach max size of cursor: %s/%s (tagNames: [ %s ])', this.counter.size, this.counter.maxCursorSize, this.metric.sortedTags.join(','));
            return;
        }
        this.counter.size++;
        return newNode(new cursor_1.Cursor({
            metric: this.metric,
            index,
        }));
    }
    get(index) {
        if (!this.root) {
            this.root = this.createNode(index);
            return this.root;
        }
        let curAxis = 0;
        let curNode = this.root;
        while (curNode) {
            if (isMatch(curNode, index)) {
                return curNode;
            }
            if (index[curAxis] >= curNode.value.index[curAxis]) {
                if (!curNode.right) {
                    return (curNode.right = this.createNode(index));
                }
                else {
                    curNode = curNode.right;
                }
            }
            else {
                if (!curNode.left) {
                    return (curNode.left = this.createNode(index));
                }
                else {
                    curNode = curNode.left;
                }
            }
            if (curAxis === this.d - 1) {
                curAxis = 0;
            }
            else {
                curAxis++;
            }
        }
    }
    walkNodesWithSize(root, visit) {
        if (!root)
            return 0;
        const lSize = this.walkNodesWithSize(root.left, visit);
        const rSize = this.walkNodesWithSize(root.right, visit);
        visit(root, lSize, rSize);
        return lSize + rSize + 1;
    }
    walkNodes(root, visit) {
        if (!root)
            return;
        this.walkNodes(root.left, visit);
        this.walkNodes(root.right, visit);
        visit(root);
    }
    findNode(index) {
        let curAxis = 0;
        let curNode = this.root;
        while (curNode) {
            if (isMatch(curNode, index)) {
                return curNode;
            }
            if (index[curAxis] >= curNode.value.index[curAxis]) {
                curNode = curNode.right;
            }
            else {
                curNode = curNode.left;
            }
            if (curAxis === this.d - 1) {
                curAxis = 0;
            }
            else {
                curAxis++;
            }
        }
    }
    buildNode(values, axis) {
        if (!values.length)
            return;
        if (values.length === 1) {
            return newNode(values[0].value);
        }
        values.sort((a, b) => (a.value.index[axis] < b.value.index[axis] ? 1 : -1));
        let mid = ~~(values.length / 2);
        let minIdx = mid - 1;
        while (minIdx >= 0) {
            if (values[minIdx].value.index[axis] != values[mid].value.index[axis]) {
                break;
            }
            mid--;
            minIdx--;
        }
        const root = values[mid];
        const nextAxis = (axis + 1) % this.d;
        const left = this.buildNode(values.slice(0, mid), nextAxis);
        const right = this.buildNode(values.slice(mid + 1), nextAxis);
        return newNode(root.value, left, right);
    }
    gc(opt) {
        let totalCount = 0;
        const expired = [];
        const nonExpired = [];
        this.walkNodes(this.root, node => {
            totalCount++;
            if (node.value.isExpired || node.value.gc(opt)) {
                expired.push(node);
            }
            else {
                nonExpired.push(node);
            }
        });
        if (!expired.length || expired.length / totalCount < opt.threshold) {
            return;
        }
        const newRoot = this.buildNode(nonExpired, 0);
        this.root = newRoot;
        this.counter.size -= expired.length;
    }
}
exports.KdTree = KdTree;
class KdTreeCollector extends collector_1.Collector {
    constructor() {
        super(...arguments);
        this.sharding = new Map();
        this.maxCursorSize = consts_1.defaultMaxCursorSize;
    }
    get(opt) {
        const shardIndex = (0, util_1.simpleHash)(opt.index) % consts_1.defaultShardingSize;
        let tree = this.sharding.get(shardIndex);
        if (!tree) {
            this.sharding.set(shardIndex, (tree = new KdTree(this.metric, this)));
        }
        const node = tree.get(opt.index);
        if (!node) {
            return null;
        }
        return node.value;
    }
    gc(opt) {
        for (const [_, tree] of this.sharding) {
            tree.gc(opt);
        }
    }
    each(cb) {
        for (const [_, tree] of this.sharding) {
            tree.walkNodes(tree.root, node => cb(node.value));
        }
    }
}
exports.KdTreeCollector = KdTreeCollector;
