// @byted-nodex/metrics/lib/collector/map.js
"use strict";
/**
 * 用 Map 来动态索引序列
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MapCollector = void 0;
const collector_1 = require("./collector");
const cursor_1 = require("./cursor");
const consts_1 = require("../consts");
const util_1 = require("../util");
const assert_1 = __importDefault(require("assert"));
class MapCollector extends collector_1.Collector {
    constructor() {
        super(...arguments);
        this.tagPool = new Map();
        this.maxCursorSize = consts_1.defaultMaxCursorSize;
    }
    get(opt) {
        const tagNames = opt.tagNames;
        (0, assert_1.default)(tagNames, 'tagNames is required in map collector');
        tagNames.forEach(tagName => {
            if (!(0, util_1.isValidString)(tagName)) {
                throw new Error(`${tagName} is not a valid tag name`);
            }
        });
        const tagKeysStr = tagNames.join('_') || 'default';
        let cursorPool;
        if (!this.tagPool.has(tagKeysStr)) {
            this.tagPool.set(tagKeysStr, (cursorPool = new Map()));
        }
        else {
            cursorPool = this.tagPool.get(tagKeysStr);
        }
        const cursorIndex = opt.index.join('|');
        if (cursorPool.has(cursorIndex)) {
            return cursorPool.get(cursorIndex);
        }
        if (this.size >= this.maxCursorSize) {
            this.metric.logger.error('Reach max size of collector pool, tagPool: %s, cursorPool: %s, cursorSize: %s/%s (tagNames: [ %s ])', this.tagPool.size, cursorPool.size, this.size, this.maxCursorSize, tagNames.join(','));
            return null;
        }
        const c = new cursor_1.SlowCursor({
            index: opt.index,
            metric: this.metric,
            tagNames,
        });
        this.size++;
        cursorPool.set(cursorIndex, c);
        return c;
    }
    gc(opt) {
        this.tagPool.forEach((cursorPool, tagIndex) => {
            cursorPool.forEach((cursor, cursorIndex) => {
                if (cursor.gc(opt)) {
                    cursorPool.delete(cursorIndex);
                    this.size--;
                }
            });
            if (!cursorPool.size) {
                this.tagPool.delete(tagIndex);
            }
        });
    }
    each(cb) {
        for (const [_, cursorPool] of this.tagPool) {
            for (const [_, cursor] of cursorPool) {
                cb(cursor);
            }
        }
    }
}
exports.MapCollector = MapCollector;
