// @universe-design/react/src/react-mini-form/Form.tsx
/* eslint-disable max-depth */
import type { FormEventHandler, ReactElement, ReactNode } from 'react';
import React, { forwardRef, memo, useCallback, useImperativeHandle } from 'react';
import type { FormContextType } from './FormContext';
import type { FormErrors, Path } from './shared';
import { useForm } from './internal/useFormInternal';
import { FormInternalContextProvider } from './internal/FormInternalContext';

/**
 * @public
 */
export interface UseFormProps<Values> {
  /**
   * 传入当前 `values` 和 校验表单的 `name`，你可以基于 `name` 做一些性能优化。
   */
  validate?: (values: Values, name?: Path) => Promise<FormErrors<Values> | void> | FormErrors<Values> | void;
  /**
   * 初始化值，如果为可选值，可以标记为可选属性。
   */
  initialValues: Values;
  /**
   * 开启 `debug` 模式
   * @defaultValue false
   * @internal
   */
  debug?: boolean;
  /**
   * 通过 Validation 之后的提交过程回调。
   */
  onSubmit?: (values: Values) => Promise<void> | void;
  /**
   * 	用户交互致字段值更新时触发回调事件，受控修改时不会触发
   */
  onValuesChange?: (changedValues: Values, allValues: Values) => void;
  /**
   * 提交失败的回调
   */
  onSubmitError?: (values: FormErrors<Values>) => void;
}

export interface FormProps<Values> extends UseFormProps<Values> {
  /**
   * 表单的子节点可以是 `Form.Item` 组合的任意节点。
   */
  children?: ReactNode;
  /**
   * Form 组件的 `className`。
   */
  className?: string;
  /**
   * 是否使用 `form` 元素包裹。
   */
  component?: 'form' | false;
}

export const Form = memo(
  forwardRef<FormContextType<any>, FormProps<any>>(function MiniForm(
    { className, component: F = 'form', validate, initialValues, onSubmit, onValuesChange, children, debug },
    ref,
  ) {
    const formCtx = useForm({
      initialValues,
      validate,
      onSubmit,
      onValuesChange,
      debug,
    });
    const { submit } = formCtx;

    useImperativeHandle(ref, () => formCtx, [formCtx]);

    // ================ Handle submit ================
    const handleSubmit = useCallback<FormEventHandler<HTMLFormElement>>(
      e => {
        e.preventDefault();
        e.stopPropagation();

        submit();
      },
      [submit],
    );

    const childNode = children;

    if (!F) {
      return <FormInternalContextProvider value={formCtx}>{childNode}</FormInternalContextProvider>;
    }

    return (
      <FormInternalContextProvider value={formCtx}>
        <F className={className} onSubmit={handleSubmit}>
          {childNode}
        </F>
      </FormInternalContextProvider>
    );
  }),
) as <Values>(props: FormProps<Values> & React.RefAttributes<FormContextType<Values>>) => ReactElement | null;
