// @universe-design/react/src/react-mini-form/FormSelector.tsx
import type { ReactElement } from 'react';
import React from 'react';
import type { FormContextType } from './FormContext';
import { useFormContextSelector } from './FormContext';

export interface ContextSelector<Values, Selected> {
  (values: FormContextType<Values>): Selected;
}

type AnySelectors = Record<string, ContextSelector<any, any>>;

/**
 * @public
 * @typeParam Values - 表单值类型
 * @typeParam Selected - 选择器返回值
 *
 * 表单数据选择器，使用方法见 [基础用法](/develop/react/form#基础用法)
 */
export function FormSelector<Values, Selected>({
  selector,
  children,
}: {
  /** 选择器 */
  selector: ContextSelector<Values, Selected>;
  /** 子元素渲染函数 */
  children: (selected: Selected) => ReactElement | null;
}) {
  return children(useFormContextSelector(selector));
}

type SelectedValues<Selectors extends AnySelectors> = Selectors extends AnySelectors
  ? { [P in keyof Selectors]: ReturnType<Selectors[P]> }
  : never;

interface SelectorItem {
  key: string;
  selector: ContextSelector<any, any>;
}

function FormSelectorsArray<Selectors extends AnySelectors>({
  selectors = [],
  selectedValues = {},
  children,
}: {
  selectors: SelectorItem[];
  children: (selected: SelectedValues<Selectors>) => ReactElement | null;
  selectedValues?: Record<string, any>;
}) {
  const nextSelectors = [...selectors];
  // 当 pop 以后 nextSelectors 为空时我们不会继续递归，所以不会出现 pop 为空的情况
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const currentSelector = nextSelectors.pop()!;
  const value = useFormContextSelector(currentSelector.selector);

  const values = {
    ...selectedValues,
    [currentSelector.key]: value,
  } as SelectedValues<Selectors>;

  if (nextSelectors.length) {
    return (
      <FormSelectorsArray selectedValues={values} selectors={nextSelectors}>
        {children}
      </FormSelectorsArray>
    );
  }

  return children(values);
}

/**
 * @public
 * @typeParam Selectors - 批量选择器
 *
 * 表单数据批量选择器，使用方法见 [复杂的动态增减表单项](/develop/react/form#复杂的动态增减表单项)
 */
export function FormSelectors<Selectors extends AnySelectors>(props: {
  /** 批量选择器 */
  selectors: Selectors;
  /** 子元素渲染函数 */
  children: (selected: SelectedValues<Selectors>) => ReactElement | null;
}): ReactElement | null {
  const { selectors, children } = props;
  const selectorArray = Object.entries(selectors).map(([key, selector]) => ({
    key,
    selector,
  }));
  return <FormSelectorsArray selectors={selectorArray}>{children}</FormSelectorsArray>;
}
