// @universe-design/react/src/react-mini-form/shared.ts
import isPlainObject from 'lodash/isPlainObject';
import { isFunction } from '../util';

export type FormErrors<T> = T extends Record<any, any> ? { [P in keyof T]?: FormErrors<T[P]> | string } : never;

export type FormTouched<T> = T extends Record<any, any> ? { [P in keyof T]?: FormTouched<T[P]> | boolean } : never;
/**
 * @public
 */
export type OneLevelPath = string | number;
/**
 * @public
 */
export type MultiLevelPath = OneLevelPath[];
/**
 * @public
 */
export type Path = MultiLevelPath | OneLevelPath;
export type SerializedName = string;

export type ValidatorFunction<V = any> = (value: V, name: Path) => string | void | Promise<string | void>;

export interface FunctionValidator<V = any> extends ValidatorFunction<V> {
  /** 设置字段校验的回调函数名称 */
  validateTrigger?: InternalValidateTrigger | InternalValidateTrigger[];
}

/**
 * @public
 */
export interface ObjectValidator<V = any> {
  /** 设置字段校验的回调函数名称 */
  validateTrigger?: InternalValidateTrigger | InternalValidateTrigger[];
  /** 校验函数 */
  validate: ValidatorFunction<V>;
}

/**
 * @public
 */
export interface FinalValidator<V = any> extends Omit<ObjectValidator<V>, 'validateTrigger'> {
  validateTrigger: InternalValidateTrigger[];
}
/**
 * @public
 */
export type Validator<V = any> = FunctionValidator<V> | ObjectValidator<V>;

export const DEFAULT_VALIDATE_TRIGGER = 0x1f1f1f;
/**
 * @public
 */
export type ValidateTrigger = string;
/**
 * @public
 */
export type InternalValidateTrigger = ValidateTrigger | typeof DEFAULT_VALIDATE_TRIGGER;

export type ValidateOptions = {
  triggerName?: string;
};

/**
 * @param value 被校验的字段值
 * @param serializedName 哪个字段触发
 * @param namePaths 触发哪些字段的校验
 * @param triggerName 触发事件
 *
 */
export type ValidateEventParam = {
  value?: any;
  serializedName?: SerializedName;
  namePaths?: Path[];
  triggerName: InternalValidateTrigger;
};

export function asArray<T>(data?: T | T[]) {
  if (Array.isArray(data)) {
    return data;
  }
  if (data === undefined) {
    return [];
  }
  return [data];
}

export function normalizeValidator<V = any>(rawValidator: Validator<V>): FinalValidator<V> {
  if (isFunction(rawValidator)) {
    return {
      validate: rawValidator,
      validateTrigger: asArray(rawValidator.validateTrigger),
    };
  }
  return {
    ...rawValidator,
    validateTrigger: asArray(rawValidator.validateTrigger),
  };
}

const SEP = '/';
const TILDE = '~';
const ESCAPED_TILDE = '~0';
const ESCAPED_SEP = '~1';

export function escapeNamePath(name: OneLevelPath): string {
  const path = String(name);
  if (path.indexOf(SEP) === -1 && path.indexOf(TILDE) === -1) {
    return path;
  }
  return path.replace(new RegExp(TILDE, 'g'), ESCAPED_TILDE).replace(new RegExp(SEP, 'g'), ESCAPED_SEP);
}

export function unescapeNamePath(name: string): string {
  return name.replace(new RegExp(ESCAPED_SEP, 'g'), SEP).replace(new RegExp(ESCAPED_TILDE, 'g'), TILDE);
}

export function marshalNamePath(name: Path): SerializedName {
  return asArray(name).map(escapeNamePath).join(SEP);
}

export function unmarshalNamePath(nameStr: SerializedName): string[] {
  const parts = nameStr.split(SEP).map(unescapeNamePath);
  if (parts.length === 1 && parts[0] === '') {
    return [];
  }
  return parts;
}

export function printFormValidationError(err: any) {
  console.warn(
    '[ud: Form] Error occurs when processing validation. We use error message to determine dirty state of the form, please catch unexpected error and return or resolve error message',
    err,
  );
}

/**
 * 解决 lodash merge 跳过 undefined 的问题, 以及 fields list 的 merge 逻辑问题
 * 1. 只要是 key in obj 对应的 value，我们都会进行 merge 处理。也就是说 a: undefined 会被执行 merge，而没有 a 这个 key 的时候就跳过 a 的 merge
 * 2. 所有的 array 都要进行判断:
 *    如果当前的 path 在 pathSet 中，则说明这个 array 是一个field，这时候的操作应该是 replace([1, 2, 3], [4, 5]) = [4, 5]
 *    如果当前的 path 不在 pathSet 中，则说明这个 array 表示一个 field list，这时候进行 index 对位 merge。merge([1, 2, 3], [4, 5]) = [4, 5, 3]
 */
export function mergeWithUndefined<TObject, TSource>(
  target: TObject,
  source: TSource,
  fieldPathSet: Set<SerializedName>,
  curPath: MultiLevelPath = [],
) {
  if (!source) {
    return target;
  }
  if (!target) {
    return source;
  }

  Object.keys(source).forEach(key => {
    const value = (source as any)[key];
    const prevValue = (target as any)[key];
    const nextPath = [...curPath, key];

    // 如果该 path 属于一个 field -> replace
    if (fieldPathSet.has(marshalNamePath(nextPath))) {
      (target as any)[key] = value;
    } else {
      // 如果 prevValue 不是 plain 的则可以直接覆盖，如果 value 不是 plain 的也可以直接覆盖
      const recursive =
        (Array.isArray(prevValue) || isPlainObject(prevValue)) && (Array.isArray(value) || isPlainObject(value));
      (target as any)[key] = recursive ? mergeWithUndefined(prevValue, value, fieldPathSet, nextPath) : value;
    }
  });

  return target;
}
