// @universe-design/react/src/text/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Basic from '../demo/0.basic';
import C1Variant from '../demo/1.variant';
import C2Overflow from '../demo/2.overflow';
import C3Disabled from '../demo/3.disabled';
import C4Alignment from '../demo/4.alignment';
import C5FontVariant from '../demo/5.font-variant';
if (!(C0Basic as Story).storyName) {
  (C0Basic as Story).storyName = '0. 基础用法';
  (C0Basic as Story).parameters = {
    docs: {
      description: {
        story: '`Text` 组件默认使用 `body0` 正文排版样式和 `span` 标签，可以通过 `as` 切换标签。\n\n',
      },
    },
  };
}
if (!(C1Variant as Story).storyName) {
  (C1Variant as Story).storyName = '1. 排版样式预设';
  (C1Variant as Story).parameters = {
    docs: {
      description: {
        story:
          '可以使用 `variant` 指定排版样式，不建议嵌套 `Text` 组件。如果需要继承父级排版样式，可以设置为 `plain`。\n\n',
      },
    },
  };
}
if (!(C2Overflow as Story).storyName) {
  (C2Overflow as Story).storyName = '2. 文字溢出';
  (C2Overflow as Story).parameters = {
    docs: {
      description: {
        story:
          '`Text` 组件默认采用文字换行布局，可以使用 `lineClamp` 属性指定最多显示的行数。指定 `lineClamp` 后，文字溢出会展示省略号，并且鼠标悬停可以弹出完整的文字提示。\n\n',
      },
    },
  };
}
if (!(C3Disabled as Story).storyName) {
  (C3Disabled as Story).storyName = '3. 禁用';
  (C3Disabled as Story).parameters = {
    docs: {
      description: {
        story:
          '禁用文字状态下用户无法选中文字，并会展示 `not-allowed` 光标。默认情况下禁用状态的文字溢出依然生效，如果需要同时禁用文字溢出弹出能力则可以配合 `tooltipProps` 进行设置。\n\n',
      },
    },
  };
}
if (!(C4Alignment as Story).storyName) {
  (C4Alignment as Story).storyName = '4. 文字对齐方式';
  (C4Alignment as Story).parameters = {
    docs: {
      description: {
        story:
          '可以通过 `alignment` 属性调整文字对齐方式。为了更好的国际化支持，我们采用 `start` 和 `end` 而非 `left` 和 `right` 表示行首行尾。\n\n',
      },
    },
  };
}
if (!(C5FontVariant as Story).storyName) {
  (C5FontVariant as Story).storyName = '5. 字体变体';
  (C5FontVariant as Story).parameters = {
    docs: {
      description: {
        story: '各种字体变体。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Text/官网示例',
  tags: ['autodocs'],
};

export { C0Basic, C1Variant, C2Overflow, C3Disabled, C4Alignment, C5FontVariant };
