// @universe-design/react-utilities/lib-esm/components/checkbox/useCheckbox.js
import { useCallback } from 'react';
import clsx from 'clsx';
import { useIsomorphicLayoutEffect } from '../../hooks/useIsomorphicLayoutEffect';
import { useControlledState } from '../../hooks/useControlledState';
/**
 * checkbox 组件的 hook
 * @public
 * @param props - checkbox 的 props
 * @param inputRef - input ref
 */
export const useCheckbox = (props, inputRef) => {
  const {
    defaultChecked,
    checked,
    name,
    style,
    className,
    error,
    disabled,
    indeterminate,
    onChange,
    classPrefix
  } = props;
  const [checkedState, setCheckedState] = useControlledState(Boolean(defaultChecked), Boolean(checked), {
    isControlledOverride: 'checked' in props
  });
  const syncIndeterminate = useCallback(() => {
    if (indeterminate !== undefined && inputRef.current) {
      inputRef.current.indeterminate = indeterminate;
    }
  }, [indeterminate, inputRef]);
  useIsomorphicLayoutEffect(() => {
    syncIndeterminate();
  }, [syncIndeterminate]);
  return {
    checked: checkedState,
    indeterminate,
    inputProps: {
      name,
      disabled,
      type: 'checkbox',
      checked: checkedState,
      onChange: e => {
        syncIndeterminate();
        setCheckedState(e.target.checked, () => {
          onChange === null || onChange === void 0 ? void 0 : onChange(e.target.checked, e);
        });
      },
      role: 'checkbox',
      ['aria-checked']: function () {
        var _inputRef$current;
        if (indeterminate || (_inputRef$current = inputRef.current) !== null && _inputRef$current !== void 0 && _inputRef$current.indeterminate) {
          return 'mixed';
        }
        return checkedState;
      }(),
      className: clsx(`${classPrefix}__input`, disabled && `${classPrefix}__input-disabled`)
    },
    wrapperProps: {
      style,
      className: clsx(`${classPrefix}__wrapper`, {
        [clsx(`${classPrefix}--error`)]: error,
        [clsx(`${classPrefix}--disabled`)]: disabled
      }, className)
    },
    descriptionProps: {
      className: clsx(`${classPrefix}__label-content`, disabled && `${classPrefix}__label-content-disabled`)
    }
  };
};
