// @universe-design/react/es/shared/components/transition/SlideTransition.js
/* eslint-disable @typescript-eslint/no-unused-vars */

import { createInOutTransition } from '@universe-design/react-utilities';
import { transitionTimingFunctionStandard } from './timingFunctionConsts';
import { useCssScope } from '../../../config-provider/hooks';
export const SlideTransition = createInOutTransition({
  name: config => `slide-${config.direction}-${config.size}`,
  styles: config => {
    const translateName = ['up', 'down'].includes(config.direction) ? 'translateY' : 'translateX';
    const inStyle = `opacity: 0; transform: ${translateName}(${['right', 'down'].includes(config.direction) ? '-' : ''}${typeof config.size === 'string' ? config.size : `${config.size}px`});`;
    const outStyle = `opacity:1; transform: ${translateName}(0px);`;
    return {
      appear: inStyle,
      enter: inStyle,
      exitDone: inStyle,
      appearActive: outStyle,
      enterActive: outStyle,
      exit: outStyle,
      exitActive: inStyle
    };
  },
  transitions: {
    appearActive: t => `opacity ${t}ms ${transitionTimingFunctionStandard}, transform ${t}ms ${transitionTimingFunctionStandard}`,
    enterActive: t => `opacity ${t}ms ${transitionTimingFunctionStandard}, transform ${t}ms ${transitionTimingFunctionStandard}`,
    exitActive: t => `opacity ${t}ms ${transitionTimingFunctionStandard}, transform ${t}ms ${transitionTimingFunctionStandard}`
  },
  useCssScope
});
