// @universe-design/react/es/shared/components/transition/ZoomTransition.js
/* eslint-disable @typescript-eslint/no-unused-vars */

import { createInOutTransition } from '@universe-design/react-utilities';
import { transitionTimingFunctionStandard } from './timingFunctionConsts';
import { useCssScope } from '../../../config-provider/hooks';

/**
 * 用于缩放场景的动画，比如 tooltip popover popconfirm
 */
export const ZoomTransition = createInOutTransition({
  name: 'zoom',
  styles: {
    appear: 'opacity: 0;transform: scale(.7);',
    enter: 'opacity: 0;transform: scale(.7);',
    exitDone: 'opacity: 0;transform: scale(.7);',
    appearActive: 'opacity: 1;transform: scale(1);',
    enterActive: 'opacity: 1;transform: scale(1);',
    appearDone: 'opacity: 1;transform: scale(1);',
    enterDone: 'opacity: 1;transform: scale(1);',
    exit: 'opacity: 1;transform: scale(1);',
    exitActive: 'opacity: 0;transform: scale(0.7);'
  },
  transitions: {
    appearActive: t => `opacity ${t}ms ${transitionTimingFunctionStandard}, transform ${t}ms ${transitionTimingFunctionStandard}`,
    enterActive: t => `opacity ${t}ms ${transitionTimingFunctionStandard}, transform ${t}ms ${transitionTimingFunctionStandard}`,
    exitActive: t => `opacity ${t}ms ${transitionTimingFunctionStandard}, transform ${t}ms ${transitionTimingFunctionStandard}`
  },
  useCssScope
});
