// @byted-service/bytedtrace/lib/confProvider/asyncCache.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AsyncCache = void 0;
const tslib_1 = require("tslib");
const consts = tslib_1.__importStar(require("./consts"));
const lodash_clonedeep_1 = tslib_1.__importDefault(require("lodash.clonedeep"));
const singleflight_1 = tslib_1.__importDefault(require("@byted-service/singleflight"));
const sf = new singleflight_1.default();
class AsyncCache {
    constructor(options) {
        this.cache = {};
        this.fetcher = options.fetcher;
        this.isSame = options.isSame;
        this.changeHandler = options.changeHandler;
        this.logger = options.logger;
        setInterval(() => {
            /* istanbul ignore next */
            this.refresh();
        }, consts.DEFAULT_CACHE_REFRESH_INTERVAL).unref();
    }
    setDefault(key, val) {
        this.cache[key] = val;
    }
    async get(key, force) {
        if (this.cache[key] && !force) {
            return this.cache[key];
        }
        try {
            this.cache[key] = await this.fetcher();
            return this.cache[key];
        }
        catch (e) /* istanbul ignore next */ {
            this.logger.error('invoke fetcher error: %o', e);
            return;
        }
    }
    async getOrSet(key, def) {
        if (this.cache[key]) {
            return this.cache[key];
        }
        try {
            this.cache[key] = await this.fetcher();
            return this.cache[key];
        }
        catch (e) /* istanbul ignore next */ {
            this.cache[key] = def;
            return def;
        }
    }
    dump() {
        const res = {};
        const keys = Object.keys(this.cache);
        for (let i = 0; i < keys.length; i++) {
            const key = keys[i];
            res[key] = (0, lodash_clonedeep_1.default)(this.cache[key]);
        }
        return res;
    }
    async refresh() {
        const keys = Object.keys(this.cache);
        for (let i = 0; i < keys.length; i++) {
            const key = keys[i];
            try {
                const val = await sf.do(key, async () => this.fetcher());
                if (typeof this.isSame === 'function' &&
                    !this.isSame(this.cache[key], val)) {
                    if (typeof this.changeHandler === 'function') {
                        this.cache[key] = (0, lodash_clonedeep_1.default)(val);
                        this.changeHandler(this.cache[key]);
                        continue;
                    }
                }
            }
            catch (error) /* istanbul ignore next */ {
                this.logger.error('asyncCache refresh key: %s, error: %o', key, error);
            }
        }
    }
}
exports.AsyncCache = AsyncCache;
