// @byted-service/bytedtrace/lib/confProvider/consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tccKeyLowerBoundSamplerExpireSecond = exports.tccKeySamplerConditionSamplingFilters = exports.tccKeySamplerDetailSamplingStrategies = exports.tccKeySamplerSamplingCountPerMinAdaptive = exports.tccKeySamplerSamplingRateProbability = exports.tccKeySamplerSamplingMode = exports.tccKeySamplerMaxOperationCount = exports.tccKeyCBMaxPostTraceSamplingDetailStrategies = exports.tccKeyCBMaxDyeingSamplingRatePerOperation = exports.tccKeyCBMaxPostTraceSamplingRatePerOperation = exports.tccKeyCBMaxTxnSamplingRatePerOperation = exports.tccKeyCBMaxTxnDataLen = exports.tccKeyCBMaxEventDataLen = exports.tccKeyCBMaxTagValueLen = exports.tccKeyCBMaxCustomMetricCountInSpan = exports.tccKeyCBMaxEventCountInSpan = exports.tccKeyCBMaxTagKVCountInObj = exports.tccKeyCBMaxSpanCountInTxn = exports.tccKeyGlobalRootDisableThriftStyleMetricsByComponent = exports.tccKeyGlobalDisableThriftStyleMetricsByComponent = exports.tccKeyGlobalRootLightModeByComponent = exports.tccKeyGlobalLightModeByComponent = exports.tccKeyGlobalCustomMetricsLevel = exports.tccKeyGlobalSpanMetricsLevel = exports.tccKeyGlobalEnableReportTotalSpan = exports.tccKeyGlobalEnableSampling = exports.tccKeyGlobalEnableDyedHeader = exports.tccKeyGlobalEnablePostTrace = exports.tccKeyGlobalEnableTracer = exports.tccDefaultRetryCount = exports.tccDefaultCacheTime = exports.tccDefaultGetTimeout = exports.tccServiceDefaultFramework = exports.tccServiceLanguageNodeName = exports.tccPlatformServiceName = exports.EVENT_CONFIG_CHANGE = exports.DEFAULT_CONFIGS = exports.DEFAULT_SAMPLER_CONFIG = exports.DEFAULT_CB_CONFIG = exports.DEFAULT_GLOBAL_TRACER_CONFIG = exports.DEFAULT_CACHE_REFRESH_INTERVAL = exports.SAMPLING_MODE_PROBABILITY = exports.SAMPLING_MODE_ADAPTIVE = exports.NAMESPACE_SAMPLER = exports.NAMESPACE_CB = exports.NAMESPACE_TRACER = exports.DISABLE_TCC_GEN_LOG = void 0;
const tslib_1 = require("tslib");
const consts_1 = require("../consts");
const env = tslib_1.__importStar(require("@byted-service/env"));
exports.DISABLE_TCC_GEN_LOG = env.isDocker() ? false : true;
// Config namespaces.
exports.NAMESPACE_TRACER = 'tracer';
exports.NAMESPACE_CB = 'circuit breaker';
exports.NAMESPACE_SAMPLER = 'sampler';
// Sampling modes.
exports.SAMPLING_MODE_ADAPTIVE = 'adaptive';
exports.SAMPLING_MODE_PROBABILITY = 'probability';
exports.DEFAULT_CACHE_REFRESH_INTERVAL = consts_1.IS_DEBUG ? 2 * 1000 : 60 * 1000;
exports.DEFAULT_GLOBAL_TRACER_CONFIG = {
    enableTrace: true,
    enablePostTrace: true,
    enableDyedHeader: true,
    enableSampling: true,
    enableReportTotalSpan: false,
    spanMetricsLevel: 0,
    customMetricsLevel: 0,
};
exports.DEFAULT_CB_CONFIG = {
    maxSpanCountInTxn: 1000,
    maxTagKVCountInObj: 30,
    maxEventCountInSpan: 30,
    maxCustomMetricCountInSpan: 30,
    maxTagValueLen: 255,
    maxEventDataLen: 1024,
    // golang 是 1048576
    maxTxnDataLen: 102400,
    maxTxnSamplingRatePerOperation: 100,
    maxPostTraceSamplingRatePerOperation: 1,
    maxDyeingSamplingRatePerOperation: 1,
};
exports.DEFAULT_SAMPLER_CONFIG = {
    samplerMaxOperationCount: 1000,
    defaultStrategy: {
        samplingMode: exports.SAMPLING_MODE_PROBABILITY,
        samplingRateProbability: 0.001,
        samplingCountPerMinAdaptive: 15,
    },
    lowerBoundSamplerExpireSecond: 6 * 3600,
};
exports.DEFAULT_CONFIGS = {
    [exports.NAMESPACE_TRACER]: exports.DEFAULT_GLOBAL_TRACER_CONFIG,
    [exports.NAMESPACE_CB]: exports.DEFAULT_CB_CONFIG,
    [exports.NAMESPACE_SAMPLER]: exports.DEFAULT_SAMPLER_CONFIG,
};
// event
exports.EVENT_CONFIG_CHANGE = 'configChange';
exports.tccPlatformServiceName = 'bytedance.aiops.trace';
exports.tccServiceLanguageNodeName = 'global/js';
exports.tccServiceDefaultFramework = 'default';
exports.tccDefaultGetTimeout = 2000; // ms
exports.tccDefaultCacheTime = 30000; // ms
exports.tccDefaultRetryCount = 3;
// Global getTracer tcc keys.
exports.tccKeyGlobalEnableTracer = 'global_enable_trace';
exports.tccKeyGlobalEnablePostTrace = 'global_enable_post_trace';
exports.tccKeyGlobalEnableDyedHeader = 'global_enable_dyed_header';
exports.tccKeyGlobalEnableSampling = 'global_enable_sampling';
exports.tccKeyGlobalEnableReportTotalSpan = 'global_enable_report_total_span';
exports.tccKeyGlobalSpanMetricsLevel = 'global_span_metrics_level';
exports.tccKeyGlobalCustomMetricsLevel = 'global_custom_metrics_level';
exports.tccKeyGlobalLightModeByComponent = 'global_light_mode_by_component';
exports.tccKeyGlobalRootLightModeByComponent = 'global_root_light_mode_by_component';
exports.tccKeyGlobalDisableThriftStyleMetricsByComponent = 'global_disable_thrift_style_metrics_by_component';
exports.tccKeyGlobalRootDisableThriftStyleMetricsByComponent = 'global_root_disable_thrift_style_metrics_by_component';
// Circuit breaker tcc keys.
exports.tccKeyCBMaxSpanCountInTxn = 'cb_max_span_count_in_transaction';
exports.tccKeyCBMaxTagKVCountInObj = 'cb_max_tagkv_count_in_obj';
exports.tccKeyCBMaxEventCountInSpan = 'cb_max_event_count_in_span';
exports.tccKeyCBMaxCustomMetricCountInSpan = 'cb_max_custom_metric_count_in_span';
exports.tccKeyCBMaxTagValueLen = 'cb_max_tag_value_data_length';
exports.tccKeyCBMaxEventDataLen = 'cb_max_event_data_length';
exports.tccKeyCBMaxTxnDataLen = 'cb_max_transaction_data_length';
exports.tccKeyCBMaxTxnSamplingRatePerOperation = 'cb_max_txn_sampling_rate_per_op';
exports.tccKeyCBMaxPostTraceSamplingRatePerOperation = 'cb_max_post_trace_sampling_rate_per_op';
exports.tccKeyCBMaxDyeingSamplingRatePerOperation = 'cb_max_dyeing_sampling_rate_per_op';
exports.tccKeyCBMaxPostTraceSamplingDetailStrategies = 'cb_max_post_trace_sampling_detail_strategies';
// Sampler tcc keys.
exports.tccKeySamplerMaxOperationCount = 'sampler_max_operation_count';
exports.tccKeySamplerSamplingMode = 'sampler_sampling_mode';
exports.tccKeySamplerSamplingRateProbability = 'sampler_sampling_rate_probability';
exports.tccKeySamplerSamplingCountPerMinAdaptive = 'sampler_sampling_count_per_min_adaptive';
exports.tccKeySamplerDetailSamplingStrategies = 'sampler_detail_sampling_strategies';
exports.tccKeySamplerConditionSamplingFilters = 'sampler_condition_sampling_filters';
exports.tccKeyLowerBoundSamplerExpireSecond = 'sampler_lower_bound_sampler_expire_second';
