// @byted-nodex/logger-core/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatter = exports.FileTransport = exports.ConsoleTransport = exports.Transport = exports.CoreLogger = exports.ConsoleLogger = exports.Logger = void 0;
const logger_1 = __importDefault(require("./logger/logger"));
exports.Logger = logger_1.default;
const console_logger_1 = __importDefault(require("./logger/console_logger"));
exports.ConsoleLogger = console_logger_1.default;
const core_logger_1 = __importDefault(require("./logger/core_logger"));
exports.CoreLogger = core_logger_1.default;
const transport_1 = __importDefault(require("./transport/transport"));
exports.Transport = transport_1.default;
const console_transport_1 = __importDefault(require("./transport/console_transport"));
exports.ConsoleTransport = console_transport_1.default;
const file_transport_1 = __importDefault(require("./transport/file_transport"));
exports.FileTransport = file_transport_1.default;
const formatter = __importStar(require("./util/formatter"));
exports.formatter = formatter;
