// @universe-design/react/lib/popconfirm/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Popconfirm = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _tooltip = require("../tooltip");
var _hooks = require("../config-provider/hooks");
var _button = require("../button");
var _InfoFilled = _interopRequireDefault(require("@universe-design/icons-react/InfoFilled"));
var _SucceedFilled = _interopRequireDefault(require("@universe-design/icons-react/SucceedFilled"));
var _WarningFilled = _interopRequireDefault(require("@universe-design/icons-react/WarningFilled"));
var _ErrorFilled = _interopRequireDefault(require("@universe-design/icons-react/ErrorFilled"));
var _util = require("../util");
var _getRenderPropValue = require("../util/getRenderPropValue");
var _stylex2 = require("../stylex");
require("./style");
const iconMap = {
  success: _SucceedFilled.default,
  warning: _WarningFilled.default,
  info: _InfoFilled.default,
  error: _ErrorFilled.default
};

// eslint-disable-next-line max-lines-per-function
const Popconfirm = exports.Popconfirm = /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
  const {
    title,
    content,
    icon = 'info',
    trigger = 'click',
    okText,
    cancelText,
    okButtonProps = {},
    cancelButtonProps = {},
    onCancel,
    onConfirm,
    disabled = false,
    mouseLeaveDelay = 100,
    ...restProps
  } = props;
  const locale = (0, _hooks.useLocale)();
  const cssScope = (0, _hooks.useCssScope)('popconfirm');
  const overlayCssScope = cssScope('__overlay');
  const udTriggerRef = (0, _react.useRef)(null);
  const renderIcon = () => {
    if (!icon) {
      return null;
    }
    if (typeof icon === 'string' && (icon === 'success' || icon === 'warning' || icon === 'info' || icon === 'error')) {
      const Icon = iconMap[icon];
      return /*#__PURE__*/_react.default.createElement("div", {
        className: cssScope(`-icon-${icon}`)
      }, /*#__PURE__*/_react.default.createElement(Icon, null));
    }
    return /*#__PURE__*/_react.default.createElement("div", {
      className: cssScope('-icon-customer')
    }, /*#__PURE__*/_react.default.createElement("span", {
      className: cssScope('-icon-customer-wrapper')
    }, icon));
  };
  const handleOnConfirm = e => {
    (0, _util.isFunction)(okButtonProps.onClick) && okButtonProps.onClick(e);
    (0, _util.isFunction)(onConfirm) && onConfirm(e);
    if (!e.isDefaultPrevented() && !disabled) {
      var _udTriggerRef$current;
      (_udTriggerRef$current = udTriggerRef.current) === null || _udTriggerRef$current === void 0 ? void 0 : _udTriggerRef$current.setVisible(false);
    }
  };
  const handleOnCancel = e => {
    (0, _util.isFunction)(cancelButtonProps.onClick) && cancelButtonProps.onClick(e);
    (0, _util.isFunction)(onCancel) && onCancel(e);
    if (!e.isDefaultPrevented() && !disabled) {
      var _udTriggerRef$current2;
      (_udTriggerRef$current2 = udTriggerRef.current) === null || _udTriggerRef$current2 === void 0 ? void 0 : _udTriggerRef$current2.setVisible(false);
    }
  };
  const handleKeyDown = (0, _react.useCallback)(e => {
    if (e.keyCode === _util.KeyCode.ESC && !e.defaultPrevented && !disabled) {
      var _udTriggerRef$current3;
      (_udTriggerRef$current3 = udTriggerRef.current) === null || _udTriggerRef$current3 === void 0 ? void 0 : _udTriggerRef$current3.setVisible(false);
    }
  }, [disabled]);
  (0, _react.useEffect)(() => {
    window.addEventListener('keydown', handleKeyDown);
    return () => {
      window.removeEventListener('keydown', handleKeyDown);
    };
  }, [handleKeyDown]);
  const renderButton = () => /*#__PURE__*/_react.default.createElement("div", {
    className: `${overlayCssScope}-button-group`
  }, /*#__PURE__*/_react.default.createElement(_button.Button, (0, _extends2.default)({
    size: "small"
  }, cancelButtonProps, {
    onClick: handleOnCancel
  }), cancelText || locale.components.popconfirm.cancelText), /*#__PURE__*/_react.default.createElement(_button.Button, (0, _extends2.default)({
    style: {
      marginRight: 0
    },
    type: "primary",
    size: "small"
  }, okButtonProps, {
    onClick: handleOnConfirm
  }), okText || locale.components.popconfirm.okText));
  return /*#__PURE__*/_react.default.createElement(_tooltip.Tooltip, (0, _extends2.default)({
    ref: ref,
    udTriggerRef: udTriggerRef,
    prefixCls: cssScope(),
    trigger: trigger,
    overlay: (() => {
      if (!title && !content) {
        return null;
      }
      return /*#__PURE__*/_react.default.createElement("div", {
        className: overlayCssScope
      }, renderIcon(), /*#__PURE__*/_react.default.createElement("div", null, title && /*#__PURE__*/_react.default.createElement("div", {
        className: (0, _clsx.default)(`${overlayCssScope}-title`, (0, _stylex.default)(_stylex2.cssBlockBase, _stylex2.cssTypographyHeadline))
      }, (0, _getRenderPropValue.getRenderPropValue)(title)), content && /*#__PURE__*/_react.default.createElement("div", {
        className: `${overlayCssScope}-content`
      }, (0, _getRenderPropValue.getRenderPropValue)(content)), renderButton()));
    })(),
    disabled: disabled,
    mouseLeaveDelay: mouseLeaveDelay,
    arrowHasStroke: true
  }, restProps));
});
