// @universe-design/react/src/image-viewer/components/Download.tsx
import * as React from 'react';
import clsx from 'clsx';
import DownloadOutlined from '@universe-design/icons-react/DownloadOutlined';
import { useLocaleReceiver } from '../../config-provider/hooks';
import { useImageViewerContext } from '../ImageViewerContext';
import Tooltip from '../../tooltip';
import { IconButton } from '../../button';
import type { ImageProps } from '../useImageViewerImages';
import '../style';

function downloadByLink(blobUrl: string, name: string) {
  const a = document.createElement('a');
  a.style.display = 'none';
  a.href = blobUrl;
  a.download = name;
  document.body.appendChild(a);
  a.click();
  document.body.removeChild(a);
}

const downloadBlob = async (blob: Blob, name: string) => {
  const blobUrl = window.URL.createObjectURL(blob);
  downloadByLink(blobUrl, name);
  window.URL.revokeObjectURL(blobUrl);
};

const downloadImage = async (source: string | Blob, name: string) => {
  if (source instanceof Blob) {
    await downloadBlob(source, name);
    return;
  }
  if (source.startsWith('blob:')) {
    downloadByLink(source, name);
    return;
  }

  const blob = await fetch(source).then(r => r.blob());
  await downloadBlob(blob, name);
};

/**
 * 图片查看器下载工具 hook
 *
 * @public
 */
export const useImageViewerToolbarDownload = (): {
  /**
   * 下载图片
   */
  download: () => void;
} => {
  const { images, currentIndex } = useImageViewerContext();

  const download = React.useCallback(async () => {
    await downloadImage(images[currentIndex].src, images[currentIndex].alt ?? `${new Date().getTime()}.jpg`);
  }, [currentIndex, images]);

  return {
    download,
  };
};

export interface DownloadProps {
  onDownload?: (image: ImageProps, context: { downloadImage: (source: string | Blob, name: string) => void }) => void;
}

export const Download = React.memo(({ onDownload }: DownloadProps) => {
  const locale = useLocaleReceiver('imageViewer');
  const { rootCssScope } = useImageViewerContext();
  const { images, currentIndex } = useImageViewerContext();

  const { download } = useImageViewerToolbarDownload();

  return (
    <Tooltip title={locale.download} overlayClassName={rootCssScope('__toolbar__tooltip')}>
      <IconButton
        size={20}
        icon={<DownloadOutlined />}
        className={clsx(rootCssScope('__toolbar__button'), rootCssScope('__toolbar__download'))}
        onClick={onDownload ? () => onDownload(images[currentIndex], { downloadImage }) : download}
      />
    </Tooltip>
  );
});
