// @universe-design/react/src/image-viewer/components/Pagination.tsx
/* eslint-disable react/no-unknown-property */
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx, makeStylexStyles } from '@universe-design/stylex/react';
import stylex from '@universe-design/stylex';
import * as React from 'react';
import LeftOutlined from '@universe-design/icons-react/LeftOutlined';
import RightOutlined from '@universe-design/icons-react/RightOutlined';
import clsx from 'clsx';
import { useEventCallback, KeyboardKey } from '@universe-design/react-utilities';
import { useLocaleReceiver } from '../../config-provider/hooks';
import { useImageViewerContext } from '../ImageViewerContext';
import Tooltip from '../../tooltip';
import { IconButton } from '../../button';
import { Space } from '../../space';

import '../style';

const usePaginationStylexStyles = makeStylexStyles(({ prefix }: { prefix: string }) => ({
  info: stylex.create(
    {
      display: 'flex',
      justifyContent: 'center',
      whiteSpace: 'nowrap',
    },
    `${prefix}__info`,
  ),
}));

/**
 * 图片查看器翻页工具 hook
 *
 * @public
 */
export const useImageViewerToolbarPagination = (): {
  /**
   * 上一页翻页按钮禁用状态
   */
  paginateBackwardDisabled: boolean;
  /**
   * 下一页翻页按钮禁用状态
   */
  paginateForwardDisabled: boolean;
  /**
   * 上一页
   */
  paginateBackward: () => void;
  /**
   * 下一页
   */
  paginateForward: () => void;
} => {
  const { images, setCurrentIndex, currentIndex } = useImageViewerContext();

  const paginateBackwardDisabled = currentIndex === 0;
  const paginateForwardDisabled = currentIndex === images.length - 1;

  const paginateBackward = useEventCallback(() => {
    if (currentIndex > 0) {
      setCurrentIndex(currentIndex - 1);
    }
  });

  const paginateForward = useEventCallback(() => {
    if (currentIndex < images.length - 1) {
      setCurrentIndex(currentIndex + 1);
    }
  });

  React.useEffect(() => {
    const onKeydown = (event: KeyboardEvent) => {
      if (event.key === KeyboardKey.ArrowLeft) {
        paginateBackward();
      } else if (event.key === KeyboardKey.ArrowRight) {
        paginateForward();
      }
    };

    document.addEventListener('keydown', onKeydown);
    return () => {
      document.removeEventListener('keydown', onKeydown);
    };
  }, [paginateBackward, paginateForward]);

  return {
    paginateBackwardDisabled,
    paginateForwardDisabled,
    paginateBackward,
    paginateForward,
  };
};

export const Pagination = () => {
  const locale = useLocaleReceiver('imageViewer');
  const { images, currentIndex, rootCssScope } = useImageViewerContext();
  const styles = usePaginationStylexStyles({
    prefix: rootCssScope('__pagination'),
  });
  const { paginateBackward, paginateBackwardDisabled, paginateForward, paginateForwardDisabled } =
    useImageViewerToolbarPagination();

  return (
    <Space size={12}>
      <Tooltip title={locale.prevArrow} overlayClassName={rootCssScope('__toolbar__tooltip')} wrapDisabledChildren>
        <IconButton
          size={20}
          disabled={paginateBackwardDisabled}
          icon={<LeftOutlined />}
          className={clsx(
            rootCssScope('__toolbar__button'),
            paginateBackwardDisabled && rootCssScope('__toolbar__button-disabled'),
          )}
          onClick={paginateBackward}
        />
      </Tooltip>
      <div sx={[styles.info]}>{`${currentIndex + 1} / ${images.length}`}</div>
      <Tooltip title={locale.nextArrow} wrapDisabledChildren overlayClassName={rootCssScope('__toolbar__tooltip')}>
        <IconButton
          disabled={paginateForwardDisabled}
          size={20}
          icon={<RightOutlined />}
          className={clsx(
            rootCssScope('__toolbar__button'),
            paginateForwardDisabled && rootCssScope('__toolbar__button-disabled'),
          )}
          onClick={paginateForward}
        />
      </Tooltip>
    </Space>
  );
};
