// @universe-design/react/src/image-viewer/components/Scaling.tsx
/* eslint-disable @typescript-eslint/no-magic-numbers */
/* eslint-disable react/no-unknown-property */
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx, makeStylexStyles } from '@universe-design/stylex/react';
import stylex from '@universe-design/stylex';
import * as React from 'react';
import OriginalSizeOutlined from '@universe-design/icons-react/OriginalSizeOutlined';
import ResizeOutlined from '@universe-design/icons-react/ResizeOutlined';
import ZoomInOutlined from '@universe-design/icons-react/ZoomInOutlined';
import ZoomOutOutlined from '@universe-design/icons-react/ZoomOutOutlined';
import { AMPLIFY_NUMBER, SCALING_SPEED } from '../constants';
import clsx from 'clsx';
import { useLocaleReceiver } from '../../config-provider/hooks/index';
import { KeyboardKey } from '@universe-design/react-utilities';
import { IconButton } from '../../button/index';
import { Tooltip } from '../../tooltip/index';
import { useEventCallback } from '../../hooks/index';
import { useImageViewerContext } from '../ImageViewerContext';
import { Space } from '../../space';

import '../style/index';

const useScalingStylexStyles = makeStylexStyles(({ prefix }: { prefix: string }) => ({
  info: stylex.create(
    {
      width: '40px',
      display: 'flex',
      justifyContent: 'center',
    },
    `${prefix}__info`,
  ),
}));

/**
 * 图片查看器缩放工具 hook
 *
 * @public
 */
export const useImageViewerToolbarScaling = (): {
  /**
   * 适应屏幕状态
   */
  fitToScreen: boolean;
  /**
   * 切换适应屏幕状态
   */
  toggleFitToScreen: () => void;
  /**
   * 缩小按钮禁用状态
   */
  scaleDownDisabled: boolean;
  /**
   * 放大按钮禁用状态
   */
  scaleUpDisabled: boolean;
  /**
   * 放大
   */
  scaleUp: () => void;
  /**
   * 缩小
   */
  scaleDown: () => void;
} => {
  // 从上下文获取图片查看器的缩放相关状态和方法
  const { scale, initTransform, scaleByOrigin, minScale, maxScale } = useImageViewerContext();

  // 适应屏幕状态，控制是否自动适配容器
  const [fitToScreen, setFitToScreen] = React.useState(true);

  // 切换适应屏幕状态的回调函数
  const toggleFitToScreen = useEventCallback(() => {
    const nextFitToScreen = !fitToScreen;
    setFitToScreen(!fitToScreen);
    initTransform({ fitToContainer: nextFitToScreen });
  });

  // 计算缩放按钮的禁用状态（保留两位小数避免浮点数精度问题）
  const scaleDownDisabled = Number(scale.toFixed(2)) <= minScale;
  const scaleUpDisabled = Number(scale.toFixed(2)) >= maxScale;

  // 放大操作处理函数
  const scaleUp = useEventCallback(() => {
    scaleByOrigin({
      getNextScale: currentScale => currentScale * (1 + SCALING_SPEED),
    });
  });

  // 缩小操作处理函数
  const scaleDown = useEventCallback(() => {
    scaleByOrigin({
      getNextScale: currentScale => currentScale * (1 - SCALING_SPEED),
    });
  });

  /**
   * Scaling
   * - 键盘操作（上下键）
   */
  React.useEffect(() => {
    /**
     * 上、下键缩放，基于容器中心缩放
     */
    const handleKeyDown = (event: KeyboardEvent) => {
      if (event.key === KeyboardKey.ArrowUp) {
        scaleUp();
      } else if (event.key === KeyboardKey.ArrowDown) {
        scaleDown();
      }
    };
    document.addEventListener('keydown', handleKeyDown);
    return () => {
      document.removeEventListener('keydown', handleKeyDown);
    };
  }, [scaleDown, scaleUp]);

  return {
    fitToScreen,
    toggleFitToScreen,
    scaleDownDisabled,
    scaleUpDisabled,
    scaleUp,
    scaleDown,
  };
};

export const Scaling = () => {
  const locale = useLocaleReceiver('imageViewer');
  const { scale, rootCssScope } = useImageViewerContext();
  const styles = useScalingStylexStyles({
    prefix: rootCssScope('__scaling'),
  });
  const { fitToScreen, toggleFitToScreen, scaleDownDisabled, scaleUpDisabled, scaleUp, scaleDown } =
    useImageViewerToolbarScaling();

  return (
    <Space size={12} className={rootCssScope('__scaling')}>
      <Tooltip title={locale.zoomOut} wrapDisabledChildren overlayClassName={rootCssScope('__toolbar__tooltip')}>
        <IconButton
          size={20}
          disabled={scaleDownDisabled}
          icon={<ZoomOutOutlined />}
          className={clsx(
            rootCssScope('__toolbar__button'),
            scaleDownDisabled && rootCssScope('__toolbar__button-disabled'),
          )}
          onClick={scaleDown}
        />
      </Tooltip>
      <div sx={[styles.info]}>{`${(scale * AMPLIFY_NUMBER).toFixed(0)}%`}</div>
      <Tooltip title={locale.zoomIn} wrapDisabledChildren overlayClassName={rootCssScope('__toolbar__tooltip')}>
        <IconButton
          disabled={scaleUpDisabled}
          size={20}
          icon={<ZoomInOutlined />}
          className={clsx(
            rootCssScope('__toolbar__button'),
            scaleUpDisabled && rootCssScope('__toolbar__button-disabled'),
          )}
          onClick={scaleUp}
        />
      </Tooltip>
      <Tooltip
        title={fitToScreen ? locale.originSize : locale.bestFit}
        overlayClassName={rootCssScope('__toolbar__tooltip')}
      >
        <IconButton
          size={20}
          icon={fitToScreen ? <OriginalSizeOutlined /> : <ResizeOutlined />}
          className={rootCssScope('__toolbar__button')}
          onClick={toggleFitToScreen}
        />
      </Tooltip>
    </Space>
  );
};
