// @universe-design/react/es/pin-input/usePinInput.js
//.ur7e4de{caret-color:transparent;}
import u1wa1nxx from "./../.stylex/u1wa1nxx";
//.u1ewtqcl{box-sizing:border-box;}
import u1p85sbs from "./../.stylex/u1p85sbs";
//.uses1vf{font-size:14px;}
import u19bueuo from "./../.stylex/u19bueuo";
//.u17mccla{text-align:center;}
import u1ust6hi from "./../.stylex/u1ust6hi";
//.u5ogflp{border-top-width:1px;}
import u1cpcpb5 from "./../.stylex/u1cpcpb5";
//.u1hqa2wf{border-right-width:1px;}
import u10qn3go from "./../.stylex/u10qn3go";
//.u1f09k3d{border-bottom-width:1px;}
import uflkg16 from "./../.stylex/uflkg16";
//.uinvdd3{border-left-width:1px;}
import uvn5txf from "./../.stylex/uvn5txf";
//.uzkkow9{border-top-style:solid;}
import us9j13g from "./../.stylex/us9j13g";
//.ucdblym{border-right-style:solid;}
import umt7kug from "./../.stylex/umt7kug";
//.ug706s2{border-bottom-style:solid;}
import uh3lwu4 from "./../.stylex/uh3lwu4";
//.ujik90z{border-left-style:solid;}
import ur61548 from "./../.stylex/ur61548";
//.u2hkw1w:focus-visible{outline-style:none;}
import u1ul1s2h from "./../.stylex/u1ul1s2h";
//.u2ys1bh{transition-property:border-color;}
import u7vf9al from "./../.stylex/u7vf9al";
//.u15vu5tu{transition-duration:0.1s;}
import up3vbyy from "./../.stylex/up3vbyy";
import React, { useImperativeHandle } from 'react';
import stylex from '@universe-design/stylex';
import clsx from 'clsx';
import { useEffectOnce } from '../hooks';
import { createContextHook, KeyboardKey, useControlledState, createDescendantContext, useConstant } from '@universe-design/react-utilities';
import { getSizeAbbr } from '../util/size';
import { focus, isNonNullable } from '../util';
import { composeEventHandlers } from '../util/function';
import composeRefs from '../util/composeRefs';
import { useCssScope } from '../config-provider';
import { sxBorderRadius, sxInput, sxInputWidth, sxTransitionTimingFunction } from '../stylex';
import { sxPinInputMargin } from './sx';
import './style';

/**
 * ================================
 *
 * Create context to track descendants and their indices
 *
 * ================================
 */
export const [PinInputDescendantsProvider, usePinInputDescendantsContext, usePinInputDescendants, usePinInputDescendant] = createDescendantContext();

/**
 * ================================
 *
 * Create Context
 *
 * ================================
 */

export const {
  Provider: PinInputContextProvider,
  useContext: usePinInputContext
} = createContextHook('PinInputContext', {
  strict: true
});

/**
 * ================================
 *
 * Pin Input
 *
 * ================================
 */

/**
 * @public
 */

const toCharArray = value => value.split('');
const validate = (v, type) => {
  let regex = /^[0-9]+$/;
  if (type === 'alphanumeric') {
    regex = /^[0-9a-zA-Z]+$/;
  }
  if (type === 'alphabetic') {
    regex = /^[a-zA-Z]+$/;
  }
  return regex.test(v);
};
/**
 * 用于判断当前应该在哪个 input 中输入，规则是第一个空值的 input
 */
const getShouldInputIndex = (values, descendants) => {
  for (let i = 0; i < values.length; i++) {
    if (!values[i]) {
      return i;
    }
  }
  return Math.min(values.length, descendants.count() - 1);
};
export const usePinInput = (props, ref) => {
  const {
    autoFocus = false,
    value,
    defaultValue = '',
    size = 'medium',
    onChange,
    onComplete,
    placeholder = '',
    id,
    disabled = false,
    error = false,
    mask = false,
    type = 'number'
  } = props;
  const rootCssScope = useCssScope('pin-input');
  const descendants = usePinInputDescendants();
  const [focusedIndex, setFocusedIndex] = React.useState(-1);
  const [values, setValues] = useControlledState(
  // defaultValue 需要去除所有的非法字符
  toCharArray(defaultValue).filter(v => validate(v, type)) || [],
  // value
  isNonNullable(value) ? toCharArray(value).filter(v => validate(v, type)) : [],
  // options
  {
    onChange: React.useCallback(nextValues => {
      onChange === null || onChange === void 0 ? void 0 : onChange(nextValues.join(''));
      // complete 放在 change 后面
      const isComplete = nextValues.every(Boolean) && nextValues.length === descendants.count();
      if (isComplete) {
        onComplete === null || onComplete === void 0 ? void 0 : onComplete(nextValues.join(''));
      }
    }, [descendants, onChange, onComplete]),
    isControlledOverride: 'value' in props
  });

  /**
   * 设置下一个 input 为聚焦状态
   */
  const focusNext = React.useCallback(index => {
    const next = descendants.next(index, false);
    if (next) {
      focus(next.node, {
        selectTextIfInput: false
      });
    }
  }, [descendants]);
  const focusByIndex = React.useCallback((index, options) => {
    const descendant = descendants.values()[index];
    if (descendant) {
      focus(descendant.node, {
        selectTextIfInput: false,
        ...options
      });
    }
  }, [descendants]);

  /**
   * 设置 input 的单字符 value
   */
  const setValue = React.useCallback((newValue, index) => {
    const nextValues = [...values];
    nextValues[index] = newValue;
    focusNext(index);
    setValues(nextValues);
  }, [values, focusNext, setValues]);

  /**
   * 重置 pin-input 输入框内容，并 focus 第一个 input
   */
  const clear = React.useCallback(() => {
    const nextValues = Array(descendants.count()).fill('');
    setValues(nextValues);
    const first = descendants.first();
    if (first) {
      focus(first.node, {
        selectTextIfInput: false
      });
    }
  }, [descendants, setValues]);

  /**
   * autoFocus 逻辑，只生效一次，后续的变化都不触发
   */
  useEffectOnce(() => {
    if (autoFocus) {
      // useEffect 是比 useLayoutEffect 里使用 requestAnimation 和 setTimeout 调用时机都要早
      focusByIndex(getShouldInputIndex(values, descendants));
    }
  });
  useImperativeHandle(ref, () => ({
    clear,
    focus: option => {
      focusByIndex(getShouldInputIndex(values, descendants), option);
    }
  }), [clear, descendants, focusByIndex, values]);
  return {
    descendants,
    rootCssScope,
    // props
    id,
    size,
    placeholder,
    disabled,
    error,
    mask,
    // states
    values,
    focusedIndex,
    // actions
    setValue,
    setValues,
    clear,
    setFocusedIndex,
    focusByIndex,
    type
  };
};

/**
 * ================================
 *
 * Pin Input Field
 *
 * ================================
 */

/**
 * @public
 */

const sxPinInputField = {
  root: stylex.compose({
    "props": {
      "B7ck84d": u1p85sbs,
      "Be2twd7": u19bueuo,
      "fsow6f": u1ust6hi,
      "B4j52fo": u1cpcpb5,
      "Bekrc4i": u10qn3go,
      "Bn0qgzm": uflkg16,
      "ibv6hh": uvn5txf,
      "icvyot": us9j13g,
      "vrafjx": umt7kug,
      "oivjwe": uh3lwu4,
      "wvpqe5": ur61548,
      "B486eqv": u1ul1s2h,
      "Bmy1vo4": u7vf9al,
      "B3o57yi": up3vbyy
    }
  }, sxBorderRadius.sm, sxTransitionTimingFunction.standard),
  small: stylex.compose(sxInput.small, sxInputWidth.small),
  medium: stylex.compose(sxInput.medium, sxInputWidth.medium),
  large: stylex.compose(sxInput.large, sxInputWidth.large),
  complete: {
    "props": {
      "pb3013": u1wa1nxx
    }
  }
};
export const usePinInputField = (_ref, ref) => {
  let {
    sx,
    ...props
  } = _ref;
  const {
    setValue,
    setValues,
    focusByIndex,
    setFocusedIndex,
    focusedIndex,
    mask,
    disabled,
    id,
    values,
    placeholder,
    size,
    rootCssScope,
    error,
    type
  } = usePinInputContext();
  const {
    index,
    descendants,
    register
  } = usePinInputDescendant();
  const inputRef = React.useRef(null);
  const latestValues = useConstant(() => ({
    get index() {
      var _inputRef$current;
      // https://github.com/chakra-ui/chakra-ui/blob/main/packages/descendant/src/use-descendant.ts#L59
      return (_inputRef$current = inputRef.current) !== null && _inputRef$current !== void 0 && _inputRef$current.dataset.index ? Number(inputRef.current.dataset.index) : -1;
    }
  }));
  const isComplete = Boolean(index === descendants.count() - 1 && values[index]);
  const inputProps = {
    'aria-label': 'Please enter your pin code',
    inputMode: 'numeric',
    type: mask ? 'password' : 'tel',
    id: isNonNullable(id) ? `${id}-${index}` : undefined,
    disabled,
    value: values[index] || '',
    placeholder: focusedIndex === index ? '' : placeholder,
    ref: composeRefs(ref, inputRef, register),
    autoComplete: 'off',
    tabIndex: getShouldInputIndex(values, descendants) === index ? undefined : -1,
    // TODO 由于 descendant register 不会触发其他参与项目的 rerender。这个值在 input field 动态变化的时候不一定准，例如，一个 4 个 field 当前 focus 到 index 3，此时，在不改变 focusIndex 的情况下修改 field 为 3 个，这时就完全无法 tab 到 pin-input 上
    ...props,
    className: clsx(rootCssScope('-field'), rootCssScope(`-field-${getSizeAbbr(size)}`), focusedIndex >= 0 && focusedIndex >= index && rootCssScope('-field-active'), error && rootCssScope('-field-error'), disabled && rootCssScope('-field-disabled'), stylex(...(sx !== null && sx !== void 0 ? sx : [sxPinInputField.root, sxPinInputField[size], isComplete && sxPinInputField.complete, index !== 0 && sxPinInputMargin[size]])), props.className),
    style: {
      ...(mask ? {
        fontFamily: 'Tahoma, PingFang SC, Microsoft Yahei' // https://codesandbox.io/s/infallible-feather-j4oez3?file=/src/App.tsx 在测试中发现 tahoma 是 win 和 mac 都系统自带的效果最好的字体，剩下两个是各自系统默认
      } : {}),
      ...props.style
    },
    onChange: composeEventHandlers(props.onChange, event => {
      const eventValue = event.currentTarget.value;
      // input 中的 value 被删除
      if (eventValue === '') {
        setValue('', index);
        return;
      }

      // 输入值不合法
      if (!validate(eventValue, type)) {
        return;
      }

      // 普通输入元素的时候
      if (eventValue.length === 1) {
        setValue(eventValue, index);
        return;
      }

      // 如果是最后一个元素，输入 2 个字符时需要覆盖当前值
      if (index === descendants.count() - 1) {
        setValue(eventValue[1], index);
        return;
      }

      // 复制粘贴的时候
      const nextValues = values.slice(0, index).concat(eventValue.split('')).filter((_, i) => i < descendants.count());
      focusByIndex(nextValues.length - 1);
      setValues(nextValues);
    }),
    onKeyDown: composeEventHandlers(props.onKeyDown, event => {
      if (event.key === KeyboardKey.Backspace) {
        if (event.currentTarget.value === '') {
          const preInput = descendants.prev(index, false);
          if (preInput) {
            setValue('', index - 1);
            focus(preInput.node, {
              selectTextIfInput: false
            });
          }
        } else {
          setValue('', index);
        }
      }
    }),
    onMouseDown: composeEventHandlers(props.onMouseDown, ev => {
      // 如果当前 input 不该被 focus 就 preventDefault
      const shouldInputIndex = getShouldInputIndex(values, descendants);
      if (shouldInputIndex !== index) {
        ev.preventDefault();
        focusByIndex(shouldInputIndex);
      }
    }),
    onFocusChange: (focused, ev) => {
      var _props$onFocusChange;
      (_props$onFocusChange = props.onFocusChange) === null || _props$onFocusChange === void 0 ? void 0 : _props$onFocusChange.call(props, focused, ev);
      if (focused) {
        // autoFocus 的时候 index 还没有更新，但是 dom 上的 index 已经更新了，所以需要从 dom 上取 index。
        setFocusedIndex(latestValues.index);
      }
    },
    onBlur: composeEventHandlers(props.onBlur, () => {
      setFocusedIndex(-1);
    })
  };
  return {
    inputProps
  };
};
