// @universe-design/react/src/util/datastructure/LinkedList.ts
/* eslint-disable max-depth */
export class LinkedListItem<T> {
  value: T;
  next: LinkedListItem<T> | null;
  prev: LinkedListItem<T> | null;

  constructor(val: T) {
    this.value = val;
    this.next = null;
    this.prev = null;
  }
}

export class LinkedList<T> {
  private _head: LinkedListItem<T> | null;
  private _tail: LinkedListItem<T> | null;
  private _length: number;

  constructor(values: T[] = []) {
    this._head = null;
    this._tail = null;
    this._length = 0;
    if (values.length > 0) {
      values.forEach(value => {
        this.append(value);
      });
    }
  }

  get head(): T | null {
    return this._head ? this._head.value : null;
  }

  get tail(): T | null {
    return this._tail ? this._tail.value : null;
  }

  get length(): number {
    return this._length;
  }

  *iterator(): IterableIterator<T> {
    let currentItem = this._head;

    while (currentItem) {
      yield currentItem.value;
      currentItem = currentItem.next;
    }
  }

  [Symbol.iterator]() {
    return this.iterator();
  }

  // Adds the element at the end of the linked list
  append(val: T, checkDuplicates = false): boolean {
    if (checkDuplicates && this.isDuplicate(val)) {
      return false;
    }

    const newItem = new LinkedListItem<T>(val);

    if (!this._tail) {
      this._head = newItem;
      this._tail = newItem;
    } else {
      this._tail.next = newItem;
      newItem.prev = this._tail;
      this._tail = newItem;
    }

    this._length++;
    return true;
  }

  // Add the element at the beginning of the linked list
  prepend(val: T, checkDuplicates = false): boolean {
    if (checkDuplicates && this.isDuplicate(val)) {
      return false;
    }

    const newItem = new LinkedListItem<T>(val);

    if (!this._head) {
      this._head = newItem;
      this._tail = newItem;
    } else {
      newItem.next = this._head;
      this._head.prev = newItem;
      this._head = newItem;
    }

    this._length++;
    return true;
  }

  removeHead(): T | null {
    const currentItem = this._head;

    // empty list
    if (!currentItem) {
      return null;
    }

    // single item list
    if (!this._head?.next) {
      this._head = null;
      this._tail = null;

      // full list
    } else {
      this._head.next.prev = null;
      this._head = this._head.next;
      currentItem.next = null;
      currentItem.prev = null;
    }

    this._length--;
    return currentItem.value;
  }

  removeTail(): T | null {
    const currentItem = this._tail;

    // empty list
    if (!currentItem) {
      return null;
    }

    // single item list
    if (!this._tail?.prev) {
      this._head = null;
      this._tail = null;

      // full list
    } else {
      this._tail.prev.next = null;
      this._tail = this._tail.prev;
      currentItem.next = null;
      currentItem.prev = null;
    }

    this._length--;
    return currentItem.value;
  }

  first(num: number): T[] {
    const iter = this.iterator();
    const result = [];

    const n = Math.min(num, this.length);

    for (let i = 0; i < n; i++) {
      const val = iter.next();
      result.push(val.value);
    }
    return result;
  }

  toArray(): T[] {
    return [...this];
  }

  private isDuplicate(val: T): boolean {
    const set = new Set(this.toArray());
    return set.has(val);
  }
}
