// @universe-design/react/src/pagination/Pagination.tsx
import React from 'react';
import RcPagination from './internal';
import clsx from 'clsx';
import LeftBoldOutlined from '@universe-design/icons-react/LeftBoldOutlined';
import RightBoldOutlined from '@universe-design/icons-react/RightBoldOutlined';
import UpTopOutlined from '@universe-design/icons-react/UpTopOutlined';
import MoreOutlined from '@universe-design/icons-react/MoreOutlined';
import stylex from '@universe-design/stylex';
import { PaginationSelect } from './PaginationSelect';
import { useBreakpoint } from '../grid/hooks/useBreakpoint';
import { useCssScope, useDirection, useLocaleReceiver } from '../config-provider/hooks';
import type { DefaultExtraSizeType, DefaultSizeType } from '../util';
import { warning, compatibleDevWarning, deprecatedWarning } from '../util';
import type { UsePaginationProps } from './usePagination';
import { usePaginationContext } from './PaginationContext';

import './style';

/**
 * @docgen
 */
export interface PaginationProps extends Omit<UsePaginationProps, 'onPageChange' | 'onPageSizeChange'> {
  /**
   * 页码改变的回调，参数是改变后的页码及每页条数
   */
  onChange?: (page: number, pageSize?: number) => void;
  /**
   * 是否在只有一页时隐藏分页器
   * @defaultValue false
   */
  hideOnSinglePage?: boolean;
  /**
   * 是否展示页码切换器，当 `total` 大于 50 时默认为 `true`, 小尺寸默认不显示页码切换器
   * @defaultValue true
   */
  showSizeChanger?: boolean;
  /**
   * 指定每页可以显示多少条
   * @defaultValue [`10`, `20`, `50`, `100`]
   */
  pageSizeOptions?: string[];
  /**
   * 页码切换器发生变化的回调
   */
  onShowSizeChange?: (current: number, size: number) => void;
  /**
   * 是否显示快速跳转输入框
   * @defaultValue false
   */
  showQuickJumper?: boolean | { goButton?: React.ReactNode };
  /**
   * 是否显示原生 `title` 页码提示
   * @defaultValue true
   */
  showTitle?: boolean;
  /**
   * 用于显示数据总量和当前数据顺序
   */
  showTotal?: (total: number, range: [number, number]) => React.ReactNode;
  /**
   * 当为 `small` 时，是小尺寸分页
   * @defaultValue `medium`
   */
  size?: Extract<DefaultSizeType | DefaultExtraSizeType, 'small' | 'medium'>;
  /**
   * 当 `size` 未指定时，根据屏幕宽度自动调整尺寸
   */
  responsive?: boolean;
  /**
   * UD 不提供 simple 模式
   * @deprecated
   */
  simple?: boolean;
  /**
   * `style` 属性透传
   */
  style?: React.CSSProperties;
  /**
   * `className` 属性透传
   */
  className?: string;
  // prefixCls?: string;
  // selectPrefixCls?: string;
  /**
   * 自定义元素渲染器，包括 `page` | `prev` | `next` | `jump-prev` | `jump-next` 五种元素可以替换
   */
  itemRender?: (
    page: number,
    type: 'page' | 'prev' | 'next' | 'jump-prev' | 'jump-next',
    originalElement: React.ReactElement<HTMLElement>,
  ) => React.ReactNode;
  /**
   * `role` 属性透传
   */
  role?: string;
  /**
   * 是否显示较少页面内容
   * @defaultValue false
   * @deprecated 由于逻辑调整，该属性表意能力不足，请改用 `siblingCount` 和 `boundaryCount`
   */
  showLessItems?: boolean;
}

export type PaginationPosition = 'top' | 'bottom' | 'both';

export interface PaginationConfig extends PaginationProps {
  position?: PaginationPosition;
}

const usePaginationCssScope = () => useCssScope('pagination');

const sxEllipsisIcon = stylex.create({
  color: $token['icon-n2'],
});

const sxEllipsis = stylex.create({
  cursor: 'pointer',
});

const Ellipsis = ({ position }: { position: 'start' | 'end' }) => {
  const prefixCls = usePaginationCssScope()('');
  const { ellipsisReadOnly } = usePaginationContext();
  if (ellipsisReadOnly) {
    return <MoreOutlined className={stylex(sxEllipsisIcon)} />;
  }
  return (
    <a className={clsx(`${prefixCls}-item-link`, stylex(sxEllipsis))}>
      {/* You can use transition effects in the container :) */}
      <div className={`${prefixCls}-item-container`}>
        <UpTopOutlined
          // FIXME 修改为正常的 icon
          style={{
            transform: position === 'start' ? 'rotate(-90deg)' : 'rotate(90deg)',
          }}
          className={`${prefixCls}-item-link-icon`}
        />
        <span className={`${prefixCls}-item-ellipsis`}>
          <MoreOutlined />
        </span>
      </div>
    </a>
  );
};

/**
 * @public
 */
export const Pagination: React.FC<PaginationProps> = ({
  className,
  size,
  itemRender,
  simple,
  showQuickJumper,
  showSizeChanger,
  showLessItems,
  ...restProps
}) => {
  warning(typeof simple !== 'boolean', compatibleDevWarning('Pagination', ['simple']));

  warning(
    typeof showLessItems === 'undefined',
    deprecatedWarning('Pagination', 'showLessItems', ['siblingCount', 'boundaryCount']),
  );

  const { xs } = useBreakpoint();

  const direction = useDirection();
  const prefixCls = useCssScope('pagination')('');
  const selectPrefixCls = useCssScope('select')('');
  const locale = useLocaleReceiver('pagination');

  const getIconsProps = () => {
    let prevIcon = (
      <button className={`${prefixCls}-item-link`} type="button" tabIndex={-1}>
        <LeftBoldOutlined />
      </button>
    );
    let nextIcon = (
      <button className={`${prefixCls}-item-link`} type="button" tabIndex={-1}>
        <RightBoldOutlined />
      </button>
    );
    let jumpPrevIcon = <Ellipsis position="start" />;
    let jumpNextIcon = <Ellipsis position="end" />;
    // change arrows direction in right-to-left direction
    if (direction === 'rtl') {
      [prevIcon, nextIcon] = [nextIcon, prevIcon];
      [jumpPrevIcon, jumpNextIcon] = [jumpNextIcon, jumpPrevIcon];
    }
    return {
      prevIcon,
      nextIcon,
      jumpPrevIcon,
      jumpNextIcon,
    };
  };

  const isSmall = size === 'small' || Boolean(xs && !size && restProps.responsive);
  const extendedClassName = clsx(
    {
      mini: isSmall,
      [`${prefixCls}-rtl`]: direction === 'rtl',
    },
    className,
  );

  const mergedShowQuickJumper = showQuickJumper ?? false;
  const mergedShowSizeChanger =
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    showSizeChanger ?? Boolean(!isSmall && typeof restProps.total === 'number' && restProps.total > 50);

  return (
    <RcPagination
      {...restProps}
      prefixCls={prefixCls}
      selectPrefixCls={selectPrefixCls}
      {...getIconsProps()}
      className={extendedClassName}
      selectComponentClass={PaginationSelect}
      locale={locale}
      itemRender={itemRender}
      showQuickJumper={mergedShowQuickJumper}
      showSizeChanger={mergedShowSizeChanger}
    />
  );
};
if (__DEV__) {
  Pagination.displayName = 'Pagination';
}
