// @universe-design/react/src/pagination/usePagination.tsx
import * as React from 'react';
import clamp from 'lodash/clamp';
import maxBy from 'lodash/maxBy';
import { useControlledState } from '@universe-design/react-utilities';
import { isFunction, isNonNullable } from '../util';

export type PaginationPager =
  | { type: 'prev' }
  | { type: 'jump-prev' }
  | { type: 'page'; page: number }
  | { type: 'jump-next' }
  | { type: 'next' };

/**
 * utils
 */
const getPageCount = ({ total, pageSize }: { total: number; pageSize: number }) => Math.ceil(total / pageSize);

const clampPageWithRange = ({ page, minPage, maxPage }: { page: number; minPage: number; maxPage: number }) =>
  clamp(page, minPage, maxPage);

/**
 * usePagination
 *
 * 抽象的分页数据处理逻辑
 */
export const usePagination = (props: {
  /**
   * 数据总数
   */
  total?: number;
  /**
   * 当前页码的默认值
   * @defaultValue 1
   */
  defaultCurrent?: number;
  /**
   * 当前页码
   */
  current?: number;
  /**
   * 单页条目数
   */
  pageSize?: number;
  /**
   * 单页条目数默认值
   * @defaultValue 10
   */
  defaultPageSize?: number;
  /**
   * 指定当开始省略时，前页码两侧的页码数。`< 1 ... 4 5 6 ... 9 >` 即上例中的 `4` 和 `6`
   * @defaultValue 1
   */
  siblingCount?:
    | number
    | {
        start: number;
        end: number;
      };
  /**
   * 指定始终显示的开始和结束页码数。`< 1 ... 4 5 6 ... 9 >` 即上例中的 `1` 和 `9`
   * @defaultValue 1
   */
  boundaryCount?:
    | number
    | {
        start: number;
        end: number;
      };
  /**
   * 是否禁止翻页
   * @defaultValue false
   */
  disabled?: boolean;
  /**
   * 是否显示省略号快进快退按钮。
   * - true: 默认是为 true，是 `{ readOnly: false, threshold: 2 }` 的简写。
   * - `object.readOnly`: ellipsis 是否为只读状态。点击 ellipsis 默认会跳转到上（下）一个相邻不相交的 siblingWindow，只读模式下只显示省略号无法交互。
   * - `object.threshold`: ellipsis 显示的阈值。当 boundaryWindow 和 siblingWindow 中间的页码数大于等于 `object.threshold` 时将这些页码显示为 ellipsis
   * @defaultValue true
   */
  showEllipses?: boolean | { readOnly?: boolean; threshold?: number };
  /**
   * 页面切换的回调
   */
  onPageChange?: (page: number, pageSize: number) => unknown;
  /**
   * 页面尺寸切换的回调
   */
  onPageSizeChange?: (page: number, pageSize: number) => unknown;
  /**
   * 是否对 siblingWindow 的边界做限制。
   * - true: 默认值为 true。当 siblingWindow 与 boundaryWindow 或 ellipsisWindow 相交，为了避免分页器 pager 数量变化导致布局闪动，我们会对 siblingWindow 的 start 和 end 值做限制，保证 pager 数量始终等于最大值。**注意** 这个行为是「尽力而为」，当 `showEllipses.threshold` 大于 3 时会出现无法避免的布局闪动。
   * - false: 不对 siblingWindow 的边界做限制，当前页始终在 siblingWindow 内部，siblingWindow 的 start 和 end 可能会超出 [1, pageCount] 区间。
   * @defaultValue true
   */
  enableSiblingWindowBoundary?: boolean;
}) => {
  const {
    total: totalProp = 0,
    current: currentProp,
    defaultCurrent = 1,
    pageSize: pageSizeProp,
    defaultPageSize = 10,
    siblingCount: siblingCountProp,
    boundaryCount: boundaryCountProp,
    showEllipses: showEllipsesProp = true,
    onPageChange,
    onPageSizeChange,
    disabled = false,
    enableSiblingWindowBoundary = true,
  } = props;

  /**
   * ================================================
   *
   * Total Prop
   *
   * ================================================
   */
  const total = clamp(totalProp, 0, Infinity);

  /**
   * ================================================
   *
   * Sibling Count Prop
   *
   * ================================================
   */
  const siblingCount = React.useMemo(() => {
    if (!isNonNullable(siblingCountProp)) {
      return {
        start: 1,
        end: 1,
      };
    }
    if (typeof siblingCountProp === 'number') {
      return {
        start: siblingCountProp,
        end: siblingCountProp,
      };
    }
    return siblingCountProp;
  }, [siblingCountProp]);

  /**
   * ================================================
   *
   * Boundary Count Prop
   *
   * ================================================
   */
  const boundaryCount = React.useMemo(() => {
    if (!isNonNullable(boundaryCountProp)) {
      return {
        start: 1,
        end: 1,
      };
    }
    if (typeof boundaryCountProp === 'number') {
      return {
        start: boundaryCountProp,
        end: boundaryCountProp,
      };
    }
    return boundaryCountProp;
  }, [boundaryCountProp]);

  /**
   * ================================================
   *
   * Start End Ellipsis Prop
   *
   * ================================================
   */
  const { startEllipsisCount, endEllipsisCount, ellipsisThreshold, ellipsisReadOnly } = React.useMemo(() => {
    const defaultEllipsisThreshold = 2;
    const defaultEllipsisReadOnly = false;
    if (!showEllipsesProp) {
      return {
        startEllipsisCount: 0,
        endEllipsisCount: 0,
        ellipsisThreshold: defaultEllipsisThreshold,
        ellipsisReadOnly: defaultEllipsisReadOnly,
      };
    }
    if (showEllipsesProp === true) {
      return {
        startEllipsisCount: 1,
        endEllipsisCount: 1,
        ellipsisThreshold: defaultEllipsisThreshold,
        ellipsisReadOnly: defaultEllipsisReadOnly,
      };
    }
    const { threshold = defaultEllipsisThreshold, readOnly = defaultEllipsisReadOnly } = showEllipsesProp;
    return {
      startEllipsisCount: 1,
      endEllipsisCount: 1,
      ellipsisThreshold: threshold,
      ellipsisReadOnly: readOnly,
    };
  }, [showEllipsesProp]);

  /**
   * ================================================
   *
   * PageSize PageCount
   *
   * ================================================
   */
  const clampPageSize = (pageSize: number) => clamp(pageSize, 1, Infinity);
  const [pageSize, setPageSize] = useControlledState(
    clampPageSize(defaultPageSize),
    isNonNullable(pageSizeProp) ? clampPageSize(pageSizeProp) : pageSizeProp,
  );
  const pageCount = getPageCount({ total, pageSize });

  /**
   * ================================================
   *
   * Window All & clampPage
   *
   * ================================================
   *
   * 【注意】窗口空间都是闭区间
   */
  const windowAllStart = 1;
  const windowAllEnd = pageCount;
  const clampPage = React.useCallback(
    (page: number) =>
      clampPageWithRange({
        page,
        minPage: windowAllStart,
        maxPage: windowAllEnd,
      }),
    [windowAllEnd, windowAllStart],
  );

  /**
   * ================================================
   *
   * Current
   *
   * ================================================
   */
  const [current, setCurrent] = useControlledState(
    clampPage(defaultCurrent),
    isNonNullable(currentProp) ? clampPage(currentProp) : currentProp,
  );
  const setCurrentWithCallback = React.useCallback(
    (page: number) => {
      if (disabled) {
        return;
      }
      setCurrent(page, _page => isFunction(onPageChange) && onPageChange(_page, pageSize));
    },
    [disabled, onPageChange, pageSize, setCurrent],
  );

  /**
   * ================================================
   *
   * Windows: boundaryStart, sibling, boundaryEnd
   *
   * ================================================
   *
   * 【注意】窗口空间都是闭区间
   */
  const windowBoundaryStartStart = 0;
  const windowBoundaryStartEnd = boundaryCount.start;
  const windowBoundaryEndStart = pageCount + 1 - boundaryCount.end;
  const windowBoundaryEndEnd = pageCount + 1;

  const windowSiblingSize = siblingCount.start + 1 + siblingCount.end;

  /**
   * ================================================
   *
   * Pagers
   *
   * ================================================
   */
  const getPagers = React.useCallback(
    (_current: number, { autoPadding = false }: { autoPadding?: boolean } = {}) => {
      const originWindowSiblingStart = _current - siblingCount.start;
      const originWindowSiblingEnd = _current + siblingCount.end;
      let windowSiblingStart = originWindowSiblingStart;
      let windowSiblingEnd = originWindowSiblingEnd;
      if (autoPadding) {
        const firstEllipsisPage = windowBoundaryStartEnd + ellipsisThreshold + siblingCount.start + 1;
        const beforeFirstEllipsisPage = firstEllipsisPage - 1;
        const lastEllipsisPage = windowBoundaryEndStart - ellipsisThreshold - siblingCount.end - 1;
        const afterLastEllipsisPage = lastEllipsisPage + 1;

        // 相同的时候取小的那个
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        const startMax = maxBy([beforeFirstEllipsisPage, firstEllipsisPage], page => getPagers(page).length)!;
        // 相同的时候大的那个
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        const endMax = maxBy([afterLastEllipsisPage, lastEllipsisPage], page => getPagers(page).length)!;

        windowSiblingStart = Math.max(
          Math.min(windowSiblingStart, endMax - siblingCount.start),
          startMax - siblingCount.start,
        );

        windowSiblingEnd = Math.min(Math.max(windowSiblingEnd, startMax + siblingCount.end), endMax + siblingCount.end);
      }

      const pagers: PaginationPager[] = [];

      for (let i = windowAllStart; i <= windowAllEnd; ) {
        // 判断哪些页码要展示成 pager
        if (
          // 左边界显示为页码
          (i >= windowBoundaryStartStart && i <= windowBoundaryStartEnd) ||
          // 右边界显示为页码
          (i >= windowBoundaryEndStart && i <= windowBoundaryEndEnd) ||
          // sibling 显示为页码
          (i >= windowSiblingStart && i <= windowSiblingEnd) ||
          // 如果 sibling window 和 boundary window 之间距离小于 ellipsisThreshold 则处于之间的都展示为页码
          (windowSiblingStart - windowBoundaryStartEnd - 1 < ellipsisThreshold &&
            i > windowBoundaryStartEnd &&
            i < windowSiblingStart) ||
          (windowBoundaryEndStart - windowSiblingEnd - 1 < ellipsisThreshold &&
            i > windowSiblingEnd &&
            i < windowBoundaryEndStart) ||
          // threshold = 1 时，边缘处可能会判断为显示 ellipsis
          (originWindowSiblingStart <= windowSiblingStart && i >= originWindowSiblingStart && i < windowSiblingStart) ||
          (originWindowSiblingEnd >= windowSiblingEnd && i <= originWindowSiblingEnd && i > windowSiblingEnd)
        ) {
          pagers.push({
            type: 'page',
            page: i,
          });
          i++;
          continue;
        }
        // 判断哪些页码要展示成 jump-prev
        if (i === windowBoundaryStartEnd + 1 && windowBoundaryStartEnd < windowSiblingStart) {
          if (startEllipsisCount) {
            pagers.push({
              type: 'jump-prev',
            });
          }
          // 下面的情况则直接从 windowBoundaryStartEnd + 1 跳到 windowSiblingStart，优化性能
          if (i < windowSiblingStart) {
            i = windowSiblingStart;
            continue;
          }
          i++;
          continue;
        }
        // 判断哪些页码要展示成 jump-next
        if (i === windowSiblingEnd + 1) {
          if (endEllipsisCount) {
            pagers.push({
              type: 'jump-next',
            });
          }
          // 下面的情况则直接从 windowSiblingEnd + 1 跳到 windowBoundaryEndStart，优化性能
          if (i < windowBoundaryEndStart) {
            i = windowBoundaryEndStart;
            continue;
          }
          i++;
          continue;
        }
        i++;
      }

      return pagers;
    },
    [
      ellipsisThreshold,
      endEllipsisCount,
      startEllipsisCount,
      siblingCount.end,
      siblingCount.start,
      windowAllEnd,
      windowBoundaryEndEnd,
      windowBoundaryEndStart,
      windowBoundaryStartEnd,
    ],
  );

  const pagers: PaginationPager[] = [
    {
      type: 'prev',
    },
    ...getPagers(current, {
      autoPadding: enableSiblingWindowBoundary,
    }),
    {
      type: 'next',
    },
  ];

  /**
   * ================================================
   *
   * Computed Values
   *
   * ================================================
   */
  const prevDisabled = current <= 1;
  const nextDisabled = current >= pageCount;
  const prevPage = clampPage(current - 1);
  const nextPage = clampPage(current + 1);
  const jumpPrevPage = clampPage(current - windowSiblingSize);
  const jumpNextPage = clampPage(current + windowSiblingSize);

  /**
   * ================================================
   *
   * Handlers
   *
   * ================================================
   */
  const gotoPage = React.useCallback(
    (page: number) => {
      setCurrentWithCallback(clampPage(page));
    },
    [clampPage, setCurrentWithCallback],
  );

  const handleGoPrev = React.useCallback(() => {
    setCurrentWithCallback(prevPage);
  }, [prevPage, setCurrentWithCallback]);

  const handleGoNext = React.useCallback(() => {
    setCurrentWithCallback(nextPage);
  }, [nextPage, setCurrentWithCallback]);

  const handleJumpPrev = React.useCallback(() => {
    if (ellipsisReadOnly) {
      return;
    }
    setCurrentWithCallback(jumpPrevPage);
  }, [ellipsisReadOnly, jumpPrevPage, setCurrentWithCallback]);

  const handleJumpNext = React.useCallback(() => {
    if (ellipsisReadOnly) {
      return;
    }
    setCurrentWithCallback(jumpNextPage);
  }, [ellipsisReadOnly, jumpNextPage, setCurrentWithCallback]);

  const handlePageSizeChange = React.useCallback(
    (_pageSize: number) => {
      if (disabled) {
        return;
      }
      const newCurrent = clampPageWithRange({
        page: current,
        minPage: windowAllStart,
        maxPage: getPageCount({ total, pageSize: _pageSize }),
      });
      setCurrent(newCurrent);
      setPageSize(_pageSize);
      isFunction(onPageChange) && onPageChange(newCurrent, _pageSize);
      isFunction(onPageSizeChange) && onPageSizeChange(newCurrent, _pageSize);
    },
    [current, disabled, onPageChange, onPageSizeChange, setCurrent, setPageSize, total],
  );

  return {
    total,
    current,
    pageSize,
    pagers,
    prevDisabled,
    nextDisabled,
    prevPage,
    nextPage,
    jumpPrevPage,
    jumpNextPage,
    ellipsisReadOnly,
    gotoPage,
    handleGoPrev,
    handleGoNext,
    handleJumpPrev,
    handleJumpNext,
    handlePageSizeChange,
  };
};

export type UsePaginationProps = Parameters<typeof usePagination>[0];
