// @universe-design/react/es/shared/components/AutoSizer.js
/**
 * Forked and edited from https://github.com/bvaughn/react-virtualized-auto-sizer/blob/master/src/index.js
 * Thanks for bvaughn's work!
 */
import React, { useRef, useState, useMemo } from 'react';
import { createResizeObserver } from '@universe-design/react-utilities';
import { useEventCallback, useIsomorphicLayoutEffect as useLayoutEffect } from '../../hooks';
import composeRefs from '../../util/composeRefs';
export const AutoSizer = /*#__PURE__*/React.forwardRef((props, ref) => {
  /** Props */
  const {
    defaultWidth = 0,
    defaultHeight = 0,
    disableWidth = false,
    disableHeight = false,
    className,
    style,
    onResize,
    children
  } = props;

  /** State */
  const [width, setWidth] = useState(defaultWidth);
  const [height, setHeight] = useState(defaultHeight);

  /** Refs */
  const autoSizerRef = useRef(null);
  const parentNodeRef = useRef();
  const handleResize = useEventCallback(entries => {
    const parentNode = parentNodeRef.current;
    for (const entry of entries) {
      if (entry.target !== parentNode || !parentNode) {
        continue;
      }
      const contentBoxSize = Array.isArray(entry.contentBoxSize) ? entry.contentBoxSize[0] : entry.contentBoxSize; // firefox 不是数组，safari < 15.4 没有 contentBoxSize

      const size = (() => {
        if (contentBoxSize) {
          return {
            width: contentBoxSize.inlineSize,
            height: contentBoxSize.blockSize
          };
        }
        const _width = parentNode.offsetWidth;
        const _height = parentNode.offsetHeight;
        const cstyle = getComputedStyle(parentNode);
        // TODO 暂时只支持 px
        const paddingLeft = parseInt(cstyle.paddingLeft, 10) || 0;
        const paddingRight = parseInt(cstyle.paddingRight, 10) || 0;
        const paddingTop = parseInt(cstyle.paddingTop, 10) || 0;
        const paddingBottom = parseInt(cstyle.paddingBottom, 10) || 0;
        return {
          width: _width - paddingLeft - paddingRight,
          height: _height - paddingTop - paddingBottom
        };
      })();
      const widthChanged = !disableWidth && width !== size.width;
      const heightChanged = !disableHeight && height !== size.height;
      if (widthChanged) {
        setWidth(size.width);
      }
      if (heightChanged) {
        setHeight(size.height);
      }
      if (widthChanged || heightChanged) {
        onResize === null || onResize === void 0 ? void 0 : onResize(size);
      }
    }
  });
  useLayoutEffect(() => {
    var _parentNode$ownerDocu, _parentNode$ownerDocu2;
    const autoSizer = autoSizerRef.current;
    const parentNode = autoSizer === null || autoSizer === void 0 ? void 0 : autoSizer.parentNode;
    const HTMLElement = parentNode === null || parentNode === void 0 ? void 0 : (_parentNode$ownerDocu = parentNode.ownerDocument) === null || _parentNode$ownerDocu === void 0 ? void 0 : (_parentNode$ownerDocu2 = _parentNode$ownerDocu.defaultView) === null || _parentNode$ownerDocu2 === void 0 ? void 0 : _parentNode$ownerDocu2.HTMLElement;
    if (!parentNode || !HTMLElement || !(parentNode instanceof HTMLElement)) {
      return;
    }
    parentNodeRef.current = parentNode;
    const ro = createResizeObserver();
    ro.subscribe(handleResize);
    ro.observer.observe(parentNode, {
      box: 'content-box'
    });
    return () => {
      ro.unsubscribe(handleResize);
      ro.observer.disconnect();
    };
  }, [handleResize]);
  return /*#__PURE__*/React.createElement("div", {
    className: className,
    style: {
      overflow: 'visible',
      ...style
    },
    ref: useMemo(() => composeRefs(ref, autoSizerRef), [ref])
  }, children({
    width,
    height
  }));
});
