// @universe-design/react/es/shared/components/FirstLineAlign.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.u1guyajx{align-items:baseline;}
import u1gmpn6f from "./../../.stylex/u1gmpn6f";
//.u1jhi6b8{flex-grow:0;}
import ug24l6d from "./../../.stylex/ug24l6d";
//.ui64zpg{flex-shrink:0;}
import uidkx17 from "./../../.stylex/uidkx17";
//.u1rmlqtg{flex-basis:auto;}
import uotlknz from "./../../.stylex/uotlknz";
//.u1j7mq28{width:0px;}
import u1k2iffh from "./../../.stylex/u1k2iffh";
//.u1a3p1vp{overflow:hidden;}
import ux645fm from "./../../.stylex/ux645fm";
//.u14t3ns0{display:inline-block;}
import u8wgnjj from "./../../.stylex/u8wgnjj";
/** @jsxRuntime classic */
/** @jsx jsx */
import stylex from '@universe-design/stylex';
import clsx from 'clsx';
import * as React from 'react';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx, useMergeComponentProps } from '@universe-design/stylex/react';
import { cssFlexBase } from '../../stylex';
import { forwardHTMLRef } from '../../util/forwardRef';
import { useCssScope } from '../../config-provider';

/**
 * 用于撑开 flex 父级
 */
export const EmptyElement = forwardHTMLRef((props, ref) => jsx("span", _extends({
  ref: ref
}, useMergeComponentProps(props, {
  className: useCssScope('empty-inline-element')(''),
  sx: [{
    "props": {
      "a9b677": u1k2iffh,
      "Bpg54ce": ux645fm,
      "mc9l5x": u8wgnjj
    }
  }]
})), '\xa0'));
if (process.env.NODE_ENV !== "production") {
  EmptyElement.displayName = 'EmptyElement';
}

/**
 * 用于首行对齐
 */
export const FirstLineAlign = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    beforeProps,
    children,
    className,
    sx = [],
    ...restProps
  } = _ref;
  const before = (() => {
    if (!beforeProps) {
      return null;
    }
    const {
      className: beforeClassName,
      children: beforeChildren,
      sx: beforeSx = [],
      ...beforeRestProps
    } = beforeProps;
    return jsx("div", _extends({
      className: clsx(beforeClassName, stylex(cssFlexBase, {
        "props": {
          "Bt984gj": u1gmpn6f,
          "Bh6795r": ug24l6d,
          "Bnnss6s": uidkx17,
          "xawz": uotlknz
        }
      }, ...beforeSx))
    }, beforeRestProps), beforeChildren, jsx(EmptyElement, null));
  })();
  return jsx("div", _extends({
    className: clsx(className, stylex(cssFlexBase, {
      "props": {
        "Bt984gj": u1gmpn6f
      }
    }, ...sx)),
    ref: ref
  }, restProps), before, children, jsx(EmptyElement, null));
});
