// @universe-design/react/es/shared/components/FocusLock.js
import _extends from "@babel/runtime/helpers/esm/extends";
import * as React from 'react';
import ReactFocusLock from 'react-focus-lock';
import { getElementFromRefOrVirtualElement } from '../../hooks/useComplementEvent';
import { focus } from '../../util';
import { contains } from '../../util/dom';
import { isClient } from '@universe-design/react-utilities';
function deferAction(action) {
  // Hidding setImmediate from Webpack to avoid inserting polyfill
  const {
    setImmediate
  } = window;
  if (typeof setImmediate !== 'undefined') {
    setImmediate(action);
  } else {
    setTimeout(action, 1);
  }
}
const restoreFocusOptions = {
  preventScroll: true
}; // 暂时固定使用 preventScroll: true，考虑到 popper 或 no-mask dialog 在弹层打开时背景内容会滚动，导致 reference 离开可视区域，关闭时 returnFocus 又会导致 reference scroll 到可视区域让页面跳动

// https://bytedance.feishu.cn/docs/doccnZGFVplrobNF9Wt3Kzp0Dwg
// https://bytedance.feishu.cn/docx/doxcnMDgZlPUjEjM4RSuWGVTKmd
export const FocusLock = /*#__PURE__*/React.forwardRef((props, ref) => {
  const {
    initialFocusRef,
    finalFocusRef,
    restoreFocus,
    children,
    isDisabled: isDisabledProp,
    autoFocus,
    persistentFocus,
    lockFocusAcrossFrames,
    enableFocusLock,
    shards: shardsProps,
    ...restProps
  } = props;
  const onActivation = React.useCallback(function (lockArea) {
    // console.log('onActivation', {});
    // 解决 returnFocus 与 onActivation 冲突问题
    if (
    // 第一个分支判断是为了兜住嵌套 FocusLock 使用默认 returnFocus 逻辑时不要被 initialFocus 覆盖，这个逻辑不能放在 defer 里面，因为要判断的是同步的 returnFocus 结果
    [lockArea].every(shard => !contains(getElementFromRefOrVirtualElement(shard), document.activeElement)) && initialFocusRef // 如果有 initialFocusRef 的话做 initialFocus 逻辑
    ) {
      focus(initialFocusRef);
    }

    // 抵消 react 的 restoreSelection
    const {
      activeElement
    } = document;
    activeElement && deferAction(function () {
      focus(activeElement);
    });
  }, [initialFocusRef]);
  const onDeactivation = React.useCallback(function () {
    if (finalFocusRef) {
      // console.log('onDeactivation', {
      //   finalFocusRef: (() => {
      //     const { textContent, classList, id, tagName } = getThunkElement(
      //       finalFocusRef
      //     ) as HTMLElement;
      //     return `${tagName}#${id}.${[...classList].join(
      //       '.'
      //     )} | ${textContent}`;
      //   })(),
      // });
      // 这里 finalFocus 可能是 null，因为 onDeactivation 是在 didUpdate 里调用的，不能保证一定能取到兄弟节点的 ref 值
      focus(finalFocusRef, restoreFocusOptions);
      // 保证没有后续 onActivation 的时候也能抵消 react 的 restoreSelection
      deferAction(function () {
        focus(finalFocusRef, restoreFocusOptions);
      });
    }
  }, [finalFocusRef]);

  // 使用 shards 达到关闭 focusLock 的效果，同时保留 returnFocus 和 initialFocus 的逻辑
  const shards = (() => {
    if (enableFocusLock || !isClient()) {
      return shardsProps;
    }
    return [document.body, ...(shardsProps !== null && shardsProps !== void 0 ? shardsProps : [])];
  })();
  return /*#__PURE__*/React.createElement(ReactFocusLock, _extends({
    ref: ref,
    crossFrame: lockFocusAcrossFrames,
    persistentFocus: persistentFocus,
    autoFocus: autoFocus,
    disabled: isDisabledProp || initialFocusRef === null && enableFocusLock === false && restoreFocus === false,
    onActivation: onActivation,
    returnFocus: !finalFocusRef && restoreFocus ? restoreFocusOptions : false,
    shards: shards,
    onDeactivation: onDeactivation
  }, restProps), children);
});
if (process.env.NODE_ENV !== "production") {
  FocusLock.displayName = 'FocusLock';
}
