// @universe-design/react/es/shared/components/Image.js
//.u13uwng7{object-position:left top;}
import u1g072z0 from "./../../.stylex/u1g072z0";
//.u1ps3kmd{object-fit:cover;}
import uqyag7z from "./../../.stylex/uqyag7z";
//.u1kle4es{object-fit:contain;}
import um7gkow from "./../../.stylex/um7gkow";
//.u1plgu50{object-fit:none;}
import u145lqgf from "./../../.stylex/u145lqgf";
//.u14xojzb{object-position:center;}
import u1mrxcor from "./../../.stylex/u1mrxcor";
//.u1l02sjl{height:100%;}
import udg2fzc from "./../../.stylex/udg2fzc";
//.uly5x3f{width:100%;}
import u13d3rh8 from "./../../.stylex/u13d3rh8";
//.u1ewtqcl{box-sizing:border-box;}
import u1p85sbs from "./../../.stylex/u1p85sbs";
//.u14t3ns0{display:inline-block;}
import u8wgnjj from "./../../.stylex/u8wgnjj";
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx, makeStylexStyles } from '@universe-design/stylex/react';
import * as React from 'react';
import stylex from '@universe-design/stylex';
import { useEventCallback } from '@universe-design/react-utilities';
import { composeEventHandlers } from '../../util/function';
import { forwardHTMLRef } from '../../internal';
import { useCssScope } from '../../config-provider/hooks';
const loading = 'loading';
const error = 'error';
const loaded = 'loaded';
const useImageStylexStyles = makeStylexStyles(_ref => {
  let {
    prefix
  } = _ref;
  return {
    base: {
      "props": {
        "B7ck84d": u1p85sbs,
        "mc9l5x": u8wgnjj
      },
      "alias": prefix
    },
    status: {
      [loading]: {
        "props": {},
        "alias": `${prefix}--loading`
      },
      [error]: {
        "props": {},
        "alias": `${prefix}--error`
      },
      [loaded]: {
        "props": {},
        "alias": `${prefix}--loaded`
      }
    },
    // 图片适配尺寸策略
    fit: {
      center: {
        "props": {
          "st4lth": u145lqgf,
          "Ermj5k": u1mrxcor,
          "Bqenvij": udg2fzc,
          "a9b677": u13d3rh8
        },
        "alias": `${prefix}--fit-center`
      },
      contain: {
        "props": {
          "st4lth": um7gkow,
          "Ermj5k": u1mrxcor,
          "Bqenvij": udg2fzc,
          "a9b677": u13d3rh8
        },
        "alias": `${prefix}--fit-contain`
      },
      default: {
        "props": {}
      },
      cover: {
        "props": {
          "st4lth": uqyag7z,
          "Ermj5k": u1mrxcor,
          "Bqenvij": udg2fzc,
          "a9b677": u13d3rh8
        },
        "alias": `${prefix}--fit-cover`
      },
      none: {
        "props": {
          "st4lth": u145lqgf,
          "Ermj5k": u1g072z0,
          "Bqenvij": udg2fzc,
          "a9b677": u13d3rh8
        },
        "alias": `${prefix}--fit-none`
      }
    }
  };
});
export const useImage_unstable = (props, ref) => {
  const {
    fit = 'default',
    sx,
    ...htmlProps
  } = props;
  const styles = useImageStylexStyles({
    prefix: useCssScope('image')()
  });
  const [status, setStatus] = React.useState(loading);
  React.useMemo(() => {
    props.src;
    setStatus(loading);
  }, [props.src]);
  const imageProps = {
    ...htmlProps,
    ref,
    sx: [styles.base, styles.fit[fit], styles.status[status], sx],
    onLoad: useEventCallback(composeEventHandlers(props.onLoad, () => {
      setStatus(loaded);
    })),
    onError: useEventCallback(composeEventHandlers(props.onError, () => {
      setStatus(error);
    }))
  };
  return {
    state: {
      status
    },
    slotsProps: {
      root: imageProps
    }
  };
};
export const Image_unstable = forwardHTMLRef((props, ref) => {
  const {
    slotsProps
  } = useImage_unstable(props, ref);
  return jsx("img", slotsProps.root);
});
