// @universe-design/react/src/shared/index.ts
export { Highlighter, HighlightConsumer, HighlightProvider, FocusLock, ScrollArea } from './components';
export * from './components/popper';
export type { PopperSidecarProps } from './components/popper-sidecar/shared';
export * from './components/popper-sidecar/FloatingUIPopperSidecar';
export * from './components/popper-sidecar/RcTriggerPopperSidecar';
export * from './locales';
export type { CompleteLocale, PartialLocale } from './locales';
export { useThemeMode, useDerivedThemeModeState, getSystemThemeMode } from './hooks/useThemeMode';
export type { ThemeMode } from './hooks/useThemeMode';
export type {
  TriggerActionType,
  UdTriggerRef,
  UDTriggerProps,
  UDAlignType,
  TriggerHideActionType,
  SharedTimeProps,
  UdTRiggerArrow,
  BuildInPlacements,
  TriggerShowActionType,
  CustomTimeType,
  ScrollAreaProps,
  NativeScrollAreaProps,
  VirtualScrollAreaProps,
  SharedScrollAreaProps,
} from './components';
export { importLocale } from './importLocale';
export type { UseAnimatedVisibleProps } from './components/transition/useAnimatedVisible';
export type { WithSlotShorthandValue, SlotShorthandValue } from './slot';
