// @byted-service/metainfo/lib/backward.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAllBackwardValues = exports.getBackwardValue = exports.allBackwardValuesToSend = exports.sendBackwardValuesFromMap = exports.sendBackwardValues = exports.sendBackwardValue = exports.withBackwardValuesToSend = exports.setBackwardValuesFromMap = exports.setBackwardValues = exports.setBackwardValue = exports.recvAllBackwardValues = exports.recvBackwardValue = exports.withBackwardValues = void 0;
const ctx_converter_1 = require("./ctx_converter");
/**
 * WithBackwardValues returns a new context that allows passing key-value pairs backward with `SetBackwardValue` from any derived context.
 */
function withBackwardValues(ctx) {
    ctx_converter_1.ops.init(ctx);
    if (ctx_converter_1.ops.getBwCtxKeyRecv(ctx)) {
        return ctx;
    }
    ctx_converter_1.ops.setBwCtxKeyRecv(ctx, {});
    return ctx;
}
exports.withBackwardValues = withBackwardValues;
/**
 * RecvBackwardValue gets a value associated with the given key that is set by `SetBackwardValue` or `SetBackwardValues`.
 */
function recvBackwardValue(ctx, key) {
    if (!ctx) {
        return '';
    }
    ctx_converter_1.ops.init(ctx);
    const kvs = ctx_converter_1.ops.getBwCtxKeyRecv(ctx);
    if (kvs) {
        const val = kvs[key];
        if (val) {
            return val;
        }
    }
    return '';
}
exports.recvBackwardValue = recvBackwardValue;
/**
 * RecvAllBackwardValues is the batched version of RecvBackwardValue.
 */
function recvAllBackwardValues(ctx) {
    if (!ctx) {
        return {};
    }
    ctx_converter_1.ops.init(ctx);
    const kvs = ctx_converter_1.ops.getBwCtxKeyRecv(ctx);
    if (kvs) {
        return Object.assign({}, kvs);
    }
    return {};
}
exports.recvAllBackwardValues = recvAllBackwardValues;
/**
 * SetBackwardValue sets a key value pair into the context.
 */
function setBackwardValue(ctx, key, val) {
    if (!ctx) {
        return false;
    }
    ctx_converter_1.ops.init(ctx);
    const kvs = ctx_converter_1.ops.getBwCtxKeyRecv(ctx);
    if (kvs) {
        kvs[key] = val;
        return true;
    }
    return false;
}
exports.setBackwardValue = setBackwardValue;
/**
 * SetBackwardValues is the batched version of `SetBackwardValue`.
 */
function setBackwardValues(ctx, ...kvs) {
    if (!ctx) {
        return false;
    }
    ctx_converter_1.ops.init(ctx);
    const kvs_ = ctx_converter_1.ops.getBwCtxKeyRecv(ctx);
    if (kvs_) {
        if (kvs.length > 0 && kvs.length % 2 === 0) {
            for (let i = 0; i < kvs.length; i += 2) {
                kvs_[kvs[i]] = kvs[i + 1];
            }
            return true;
        }
    }
    return false;
}
exports.setBackwardValues = setBackwardValues;
/**
 * SetBackwardValuesFromMap is the batched version of `SetBackwardValue`.
 */
function setBackwardValuesFromMap(ctx, kvs) {
    if (!ctx) {
        return false;
    }
    ctx_converter_1.ops.init(ctx);
    const kvs_ = ctx_converter_1.ops.getBwCtxKeyRecv(ctx);
    if (kvs_) {
        const ks = Object.keys(kvs);
        if (ks.length > 0) {
            for (let i = 0; i < ks.length; i++) {
                kvs_[ks[i]] = kvs[ks[i]];
            }
            return true;
        }
    }
    return false;
}
exports.setBackwardValuesFromMap = setBackwardValuesFromMap;
/**
 * WithBackwardValuesToSend returns a new context that collects key-value pairs set with `SendBackwardValue` or `SendBackwardValues` into any derived context.
 */
function withBackwardValuesToSend(ctx) {
    ctx_converter_1.ops.init(ctx);
    if (ctx_converter_1.ops.getBwCtxKeySend(ctx)) {
        return ctx;
    }
    ctx_converter_1.ops.setBwCtxKeySend(ctx, {});
    return ctx;
}
exports.withBackwardValuesToSend = withBackwardValuesToSend;
/**
 * SendBackwardValue sets a key-value pair into the context for sending to a remote endpoint.
 * Note that the values can not be retrieved with `RecvBackwardValue` from the same context.
 */
function sendBackwardValue(ctx, key, val) {
    if (!ctx) {
        return false;
    }
    ctx_converter_1.ops.init(ctx);
    const kvs = ctx_converter_1.ops.getBwCtxKeySend(ctx);
    if (kvs) {
        kvs[key] = val;
        return true;
    }
    return false;
}
exports.sendBackwardValue = sendBackwardValue;
/**
 * SendBackwardValues is the batched version of `SendBackwardValue`.
 */
function sendBackwardValues(ctx, ...kvs) {
    if (!ctx) {
        return false;
    }
    ctx_converter_1.ops.init(ctx);
    const kvs_ = ctx_converter_1.ops.getBwCtxKeySend(ctx);
    if (kvs_) {
        if (kvs.length > 0 && kvs.length % 2 === 0) {
            for (let i = 0; i < kvs.length; i += 2) {
                kvs_[kvs[i]] = kvs[i + 1];
            }
            return true;
        }
    }
    return false;
}
exports.sendBackwardValues = sendBackwardValues;
/**
 * SendBackwardValuesFromMap is the batched version of `SendBackwardValue`.
 */
function sendBackwardValuesFromMap(ctx, kvs) {
    if (!ctx) {
        return false;
    }
    ctx_converter_1.ops.init(ctx);
    const kvs_ = ctx_converter_1.ops.getBwCtxKeySend(ctx);
    if (kvs_) {
        const ks = Object.keys(kvs);
        if (ks.length > 0) {
            for (let i = 0; i < ks.length; i++) {
                kvs_[ks[i]] = kvs[ks[i]];
            }
            return true;
        }
    }
    return false;
}
exports.sendBackwardValuesFromMap = sendBackwardValuesFromMap;
/**
 * AllBackwardValuesToSend retrieves all key-values pairs set by `SendBackwardValue` or `SendBackwardValues` from the given context.
 * This function is designed for frameworks, common developers should not use it.
 */
function allBackwardValuesToSend(ctx) {
    if (!ctx) {
        return {};
    }
    ctx_converter_1.ops.init(ctx);
    const kvs = ctx_converter_1.ops.getBwCtxKeySend(ctx);
    if (kvs) {
        return Object.assign({}, kvs);
    }
    return {};
}
exports.allBackwardValuesToSend = allBackwardValuesToSend;
/**
 * GetBackwardValue is an alias to `RecvBackwardValue`.
 * Deprecated: for clarity, please use `RecvBackwardValue` instead.
 */
function getBackwardValue(ctx, key) {
    return recvBackwardValue(ctx, key);
}
exports.getBackwardValue = getBackwardValue;
/**
 * GetAllBackwardValues is an alias to RecvAllBackwardValues.
 * Deprecated: for clarity, please use `RecvAllBackwardValues` instead.
 */
function getAllBackwardValues(ctx) {
    return recvAllBackwardValues(ctx);
}
exports.getAllBackwardValues = getAllBackwardValues;
