// @byted-service/metainfo/lib/ctx_converter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.converter = exports.ops = exports.updateConverter = void 0;
const consts_1 = require("./consts");
/**
 * 由于 gopkg/metainfo 变动较为频繁，这里为了兼容内存 ctx 数据，做一层内存映射，待所有版本号 <= 1.1.4 的全部移除后即可使用该功能
 */
const defaultOps = {
    /**
     * 旧 ctx 转换为新 ctx
     */
    init(ctx) {
        //
    },
    /**
     * 新 ctx 转换为旧 ctx
     */
    getCtxKey(ctx) {
        return ctx[consts_1.ctxKey];
    },
    setCtxKey(ctx, value) {
        ctx[consts_1.ctxKey] = value;
    },
    getBwCtxKeyRecv(ctx) {
        return ctx[consts_1.bwCtxKeyRecv];
    },
    setBwCtxKeyRecv(ctx, value) {
        ctx[consts_1.bwCtxKeyRecv] = value;
    },
    getBwCtxKeySend(ctx) {
        return ctx[consts_1.bwCtxKeySend];
    },
    setBwCtxKeySend(ctx, value) {
        ctx[consts_1.bwCtxKeySend] = value;
    },
    del(ctx) {
        delete ctx[consts_1.ctxKey];
        delete ctx[consts_1.ctxVersionKey];
        delete ctx[consts_1.bwCtxKey];
        delete ctx[consts_1.bwCtxKeyRecv];
        delete ctx[consts_1.bwCtxKeySend];
        delete ctx[consts_1.bwCtxKeySend];
    },
};
const defaultCtxConverter = {
    version: consts_1.version,
    ctxVersion: consts_1.ctxVersion,
    versions: {
        [consts_1.ctxVersion]: defaultOps,
    },
};
const getConverter = () => {
    let converter = (global.MetaInfo_CtxConverter =
        global.MetaInfo_CtxConverter ||
            defaultCtxConverter);
    if (converter.ctxVersion < defaultCtxConverter.ctxVersion) {
        converter = global.MetaInfo_CtxConverter = defaultCtxConverter;
    }
    if (!converter.versions[consts_1.ctxVersion]) {
        throw new Error('[metainfo] version "' +
            consts_1.version +
            '" cannot run at the same time with version "' +
            converter.version +
            '". Please update all @byted-service/metainfo to the latest');
    }
    return converter;
};
const updateConverter = () => {
    exports.converter = converter = getConverter();
    exports.ops = converter.versions[consts_1.ctxVersion];
    if (converter.ctxVersion !== defaultCtxConverter.ctxVersion) {
        const init = exports.ops.init;
        exports.ops.init = (ctx) => {
            ctx[consts_1.ctxVersionKey] = ctx[consts_1.ctxVersionKey] || consts_1.ctxVersion;
            init(ctx);
        };
    }
};
exports.updateConverter = updateConverter;
let converter = getConverter();
exports.converter = converter;
exports.ops = converter.versions[consts_1.ctxVersion];
(0, exports.updateConverter)();
/**
 * 在启动后的 1min 内，每 100ms 检测一次环境变量
 */
/* istanbul ignore next */
const intervalId = setInterval(exports.updateConverter, 100).unref();
/* istanbul ignore next */
setTimeout(() => clearInterval(intervalId), 1 * 60 * 1000).unref();
