// @byted-service/metainfo/lib/deprecated.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.saveMetaInfoToNewCtx = exports.getPersistValue = exports.getTransitValue = exports.savePersistValue = exports.saveTransitValue = exports.setUpstreamValue = exports.saveMetaInfoIntoMap = exports.isInvalidStringArgs = exports.isInvalidObjectArgs = exports.InfoType = exports.UPSTREAM_META_INFO = exports.PREFIX_TRANSIT_UPSTREAM = exports.PREFIX_TRANSIT = exports.PREFIX_PERSIST = exports.ctxKeyOld = void 0;
const info_1 = require("./info");
const utils_1 = require("./utils");
const helper_1 = require("./helper");
const consts_1 = require("./consts");
/** @deprecated */
exports.ctxKeyOld = Symbol.for('K_META_INFO');
/** @deprecated use PrefixPersistent instead */
exports.PREFIX_PERSIST = consts_1.PrefixPersistent;
/** @deprecated use PrefixTransient instead */
exports.PREFIX_TRANSIT = consts_1.PrefixTransient;
/** @deprecated use PrefixTransientUpstream instead */
exports.PREFIX_TRANSIT_UPSTREAM = consts_1.PrefixTransientUpstream;
/** @deprecated */
exports.UPSTREAM_META_INFO = Symbol.for('K_UPSTREAM_META_INFO');
/** @deprecated */
var InfoType;
(function (InfoType) {
    InfoType[InfoType["InvalidType"] = consts_1.invalidType] = "InvalidType";
    InfoType[InfoType["TransientUpstreamType"] = consts_1.transientUpstreamType] = "TransientUpstreamType";
    InfoType[InfoType["TransientType"] = consts_1.transientType] = "TransientType";
    InfoType[InfoType["PersistentType"] = consts_1.persistentType] = "PersistentType";
})(InfoType = exports.InfoType || (exports.InfoType = {}));
/** @deprecated */
function isInvalidObjectArgs(...args) {
    for (const arg of args) {
        if (!arg || typeof arg !== 'object') {
            return true;
        }
    }
    return false;
}
exports.isInvalidObjectArgs = isInvalidObjectArgs;
/** @deprecated */
function isInvalidStringArgs(...args) {
    for (const arg of args) {
        if (!arg || typeof arg !== 'string' || arg.length === 0) {
            return true;
        }
    }
    return false;
}
exports.isInvalidStringArgs = isInvalidStringArgs;
/** @deprecated use saveMetaInfoToMap instead */
function saveMetaInfoIntoMap(context, map) {
    if (isInvalidObjectArgs(context)) {
        return;
    }
    (0, utils_1.saveMetaInfoToMap)(context, map);
}
exports.saveMetaInfoIntoMap = saveMetaInfoIntoMap;
/** @deprecated use setMetaInfoFromMap instead */
function setUpstreamValue(context, value) {
    if (isInvalidObjectArgs(context, value)) {
        return context;
    }
    return (0, utils_1.setMetaInfoFromMap)(context, value);
}
exports.setUpstreamValue = setUpstreamValue;
/** @deprecated use withValue instead */
function saveTransitValue(context, data) {
    if (isInvalidObjectArgs(context, data)) {
        return context;
    }
    for (const k of Object.keys(data)) {
        const v = data[k];
        if (typeof v === 'string') {
            (0, info_1.withValue)(context, k, v);
        }
    }
    return context;
}
exports.saveTransitValue = saveTransitValue;
/** @deprecated use withPersistentValue instead */
function savePersistValue(context, data) {
    if (isInvalidObjectArgs(context, data)) {
        return context;
    }
    for (const k of Object.keys(data)) {
        const v = data[k];
        if (typeof v === 'string') {
            (0, info_1.withPersistentValue)(context, k, v);
        }
    }
    return context;
}
exports.savePersistValue = savePersistValue;
/** @deprecated use getValue instead */
function getTransitValue(context, key) {
    if (isInvalidObjectArgs(context) || isInvalidStringArgs(key)) {
        return '';
    }
    return (0, info_1.getValue)(context, key);
}
exports.getTransitValue = getTransitValue;
/** @deprecated use getPersistValue instead */
function getPersistValue(context, key) {
    if (isInvalidObjectArgs(context) || isInvalidStringArgs(key)) {
        return '';
    }
    return (0, info_1.getPersistentValue)(context, key);
}
exports.getPersistValue = getPersistValue;
/** @deprecated use copyMetaInfo instead */
function saveMetaInfoToNewCtx(oldContext, newContext) {
    if (isInvalidObjectArgs(oldContext, newContext)) {
        return;
    }
    (0, helper_1.copyMetaInfo)(oldContext, newContext);
}
exports.saveMetaInfoToNewCtx = saveMetaInfoToNewCtx;
