// @byted-service/metainfo/lib/helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rebuildOldMetainfo = exports.addKVToOld = exports.convertMetainfoToNew = exports.turnOnDoubleWrite = exports.turnOffDoubleWrite = exports.doubleWrite = exports.copyMetaInfo = exports.delMetaInfo = void 0;
const deprecated_1 = require("./deprecated");
const kv_1 = require("./kv");
const backward_1 = require("./backward");
const ctx_converter_1 = require("./ctx_converter");
const consts_1 = require("./consts");
function delMetaInfo(ctx) {
    if (ctx) {
        ctx_converter_1.ops.del(ctx);
        if (exports.doubleWrite) {
            delete ctx[deprecated_1.ctxKeyOld];
            delete ctx[deprecated_1.UPSTREAM_META_INFO];
        }
    }
}
exports.delMetaInfo = delMetaInfo;
function copyMetaInfo(fromCtx, toCtx) {
    const copy = {};
    let kv = (0, kv_1.getKV)(fromCtx);
    while (kv) {
        const key = `${kv.key}-${kv.mode}`;
        if (copy[key] !== true) {
            (0, kv_1.addKV)(toCtx, kv.mode, kv.key, kv.val);
            copy[key] = true;
        }
        kv = kv.pre;
    }
    const kvs = (0, backward_1.getAllBackwardValues)(fromCtx);
    if (Object.keys(kvs).length > 0) {
        (0, backward_1.withBackwardValues)(toCtx);
        const ks = Object.keys(kvs);
        for (let i = 0; i < ks.length; i++) {
            (0, backward_1.setBackwardValue)(toCtx, ks[i], kvs[ks[i]]);
        }
    }
}
exports.copyMetaInfo = copyMetaInfo;
/**
 * 双写开关，双写会造成多余的性能开销，需在后续迭代中用 bnpm 上去掉 `版本号 <= 1.0.2`
 */
exports.doubleWrite = process.env.METAINFO_DOUBLE_WRITE !== 'false';
function turnOffDoubleWrite() {
    exports.doubleWrite = false;
    clearInterval(intervalId);
    clearTimeout(timeoutId);
}
exports.turnOffDoubleWrite = turnOffDoubleWrite;
function turnOnDoubleWrite() {
    exports.doubleWrite = true;
    clearInterval(intervalId);
    clearTimeout(timeoutId);
}
exports.turnOnDoubleWrite = turnOnDoubleWrite;
/**
 * 在启动后的 1min 内，每 100ms 检测一次环境变量，如果环境变量 METAINFO_DOUBLE_WRITE = false，就关闭双写
 */
/* istanbul ignore next */
const intervalId = setInterval(() => (exports.doubleWrite =
    process.env.METAINFO_DOUBLE_WRITE === 'false' ? false : true), 100).unref();
/* istanbul ignore next */
const timeoutId = setTimeout(() => clearInterval(intervalId), 1 * 60 * 1000).unref();
const lenSym = Symbol.for('#len');
const mapSym = Symbol.for('#map');
const convertedMarkSym = Symbol.for('#convertedMark');
function convertMetainfoToNew(ctx) {
    const kvs = ctx[deprecated_1.ctxKeyOld];
    const upstreamMetaInfo = ctx[deprecated_1.UPSTREAM_META_INFO];
    const newKVs = ctx_converter_1.ops.getCtxKey(ctx);
    /**
     * 由于默认开启了双写，所以 new metainfo <= old metainfo 必定成立
     * 当符合以下任意一种情况时，都代表着需要重建 metainfo
     * 1. 初次执行
     * 2. ctxKeyOld 被老 api 修改
     * 3. UPSTREAM_META_INFO 被老 api 修改
     */
    if ((kvs &&
        Array.isArray(kvs) &&
        kvs.length > 0 &&
        kvs[lenSym] !== kvs.length) ||
        (upstreamMetaInfo && !upstreamMetaInfo[convertedMarkSym]) ||
        !newKVs) {
        rebuildOldMetainfo(ctx);
        const newKVs = ctx[deprecated_1.ctxKeyOld];
        let pre = null;
        for (let i = 0; i < newKVs.length; i++) {
            const { key, value, type } = newKVs[i];
            pre = {
                pre,
                mode: type,
                key,
                val: value,
            };
        }
        ctx_converter_1.ops.setCtxKey(ctx, pre);
    }
    return ctx;
}
exports.convertMetainfoToNew = convertMetainfoToNew;
function addKVToOld(ctx, type, key, value) {
    const kvs = (ctx[deprecated_1.ctxKeyOld] = ctx[deprecated_1.ctxKeyOld] || []);
    const upstreamMetaInfo = (ctx[deprecated_1.UPSTREAM_META_INFO] = (ctx[deprecated_1.UPSTREAM_META_INFO] || {}));
    const keyMap = (kvs[mapSym] = (kvs[mapSym] || {}));
    kvs[lenSym] = kvs[lenSym] || 0;
    upstreamMetaInfo[convertedMarkSym] =
        upstreamMetaInfo[convertedMarkSym] || true;
    const mapKey = `${key}-${type}`;
    if (keyMap[mapKey]) {
        kvs[keyMap[mapKey].index] = keyMap[mapKey] = {
            type,
            key,
            value,
            index: keyMap[mapKey].index,
        };
    }
    else {
        kvs.push((keyMap[mapKey] = {
            type,
            key,
            value,
            index: kvs.length,
        }));
        kvs[lenSym]++;
    }
    switch (type) {
        case consts_1.persistentType:
            upstreamMetaInfo[consts_1.PrefixPersistent + key] = value;
            break;
        case consts_1.transientType:
            upstreamMetaInfo[consts_1.PrefixTransient + key] = value;
            break;
    }
    return ctx;
}
exports.addKVToOld = addKVToOld;
/**
 * 重建的目的是为了保证 ctxKeyOld、UPSTREAM_META_INFO 两者数据一致
 */
function rebuildOldMetainfo(ctx) {
    const kvs = (ctx[deprecated_1.ctxKeyOld] = (ctx[deprecated_1.ctxKeyOld] || []));
    const upstreamMetaInfo = (ctx[deprecated_1.UPSTREAM_META_INFO] = (ctx[deprecated_1.UPSTREAM_META_INFO] || {}));
    const newKVs = [];
    const newUpstreamMetaInfo = {};
    const newKeyMap = {};
    /**
     * 由于外部使用时总是先调用 MetaInfo.setUpstreamValue，再调用 MetaInfo.setMetaInfoFromMap
     * 导致 kvs 总是比 upstreamMetaInfo 新，所以就先处理 kvs 再处理 upstreamMetaInfo
     */
    const del = {};
    for (let i = kvs.length - 1; i >= 0; i--) {
        const { key, type, value } = kvs[i];
        const mapKey = `${key}-${type}`;
        if (newKeyMap[mapKey] === undefined) {
            newKeyMap[mapKey] = {
                key,
                type,
                value,
            };
            if (typeof value !== 'string' || value.length === 0) {
                del[mapKey] = true;
            }
        }
    }
    const upstreamMetaInfoKeys = Object.keys(upstreamMetaInfo);
    for (let i = 0; i < upstreamMetaInfoKeys.length; i++) {
        let key = upstreamMetaInfoKeys[i];
        const value = upstreamMetaInfo[key];
        let type;
        switch (true) {
            case key.startsWith(consts_1.PrefixPersistent):
                key = key.slice(consts_1.PrefixPersistent.length);
                type = consts_1.persistentType;
                break;
            case key.startsWith(consts_1.PrefixTransient):
                key = key.slice(consts_1.PrefixTransient.length);
                type = consts_1.transientUpstreamType;
                break;
            default:
                key = '';
                type = consts_1.invalidType;
                break;
        }
        if (key.length > 0) {
            const mapKey = `${key}-${type}`;
            if (newKeyMap[mapKey] === undefined) {
                newKeyMap[mapKey] = {
                    key,
                    type,
                    value,
                };
                if (typeof value !== 'string' || value.length === 0) {
                    del[mapKey] = true;
                }
            }
        }
    }
    const newKeyMapKeys = Object.keys(newKeyMap);
    for (let i = 0; i < newKeyMapKeys.length; i++) {
        const key = newKeyMapKeys[i];
        if (!del[key]) {
            const kv = newKeyMap[key];
            kv.index = newKVs.length;
            newKVs.push(kv);
            switch (kv.type) {
                case consts_1.persistentType:
                    newUpstreamMetaInfo[consts_1.PrefixPersistent + kv.key] = kv.value;
                    break;
                case consts_1.transientType:
                    newUpstreamMetaInfo[consts_1.PrefixTransient + kv.key] = kv.value;
                    break;
                default:
                    break;
            }
        }
    }
    newKVs[mapSym] = newKeyMap;
    newKVs[lenSym] = newKVs.length;
    newUpstreamMetaInfo[convertedMarkSym] = true;
    ctx[deprecated_1.ctxKeyOld] = newKVs;
    ctx[deprecated_1.UPSTREAM_META_INFO] = newUpstreamMetaInfo;
}
exports.rebuildOldMetainfo = rebuildOldMetainfo;
