// @byted-service/metainfo/lib/info.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.delPersistentValue = exports.withPersistentValue = exports.getAllPersistentValues = exports.getPersistentValue = exports.delValue = exports.withValue = exports.getAllValues = exports.getValue = exports.transferForward = void 0;
const kv_1 = require("./kv");
const consts_1 = require("./consts");
// **Using empty string as key or value is not support.**
/**
 * TransferForward converts transient values to transient-upstream values and filters out original transient-upstream values.
 * It should be used before the context is passing from server to client.
 */
function transferForward(ctx) {
    if (!ctx) {
        return ctx;
    }
    return (0, kv_1.copyLink)(ctx, (mode, key, val) => {
        if (mode === consts_1.transientUpstreamType) {
            return null;
        }
        if (mode === consts_1.transientType) {
            mode = consts_1.transientUpstreamType;
        }
        return {
            mode,
            key,
            val,
            pre: null,
        };
    });
}
exports.transferForward = transferForward;
/**
 * GetValue retrieves the value set into the context by the given key.
 */
function getValue(ctx, k) {
    return (0, kv_1.getV)(ctx, consts_1.transientUpstreamType | consts_1.transientType, k);
}
exports.getValue = getValue;
/**
 * GetAllValues retrieves all transient values
 */
function getAllValues(ctx) {
    if (!ctx) {
        return {};
    }
    return (0, kv_1.getAll)(ctx, consts_1.transientUpstreamType | consts_1.transientType);
}
exports.getAllValues = getAllValues;
/**
 * WithValue sets the value into the context by the given key.
 * This value will be propagated to the next service/endpoint through an RPC call.
 *
 * Notice that it will not propagate any further beyond the next service/endpoint,
 * Use WithPersistentValue if you want to pass a key/value pair all the way.
 */
function withValue(ctx, k, v) {
    if (k.length === 0 || v.length === 0) {
        return ctx;
    }
    return (0, kv_1.addKV)(ctx, consts_1.transientType, k, v);
}
exports.withValue = withValue;
/**
 * DelValue deletes a key/value from the current context.
 * Since empty string value is not valid, we could just set the value to be empty.
 */
function delValue(ctx, k) {
    if (k.length === 0) {
        return ctx;
    }
    return (0, kv_1.addKV)(ctx, consts_1.transientType, k, '');
}
exports.delValue = delValue;
/**
 * GetPersistentValue retrieves the persistent value set into the context by the given key.
 */
function getPersistentValue(ctx, k) {
    return (0, kv_1.getV)(ctx, consts_1.persistentType, k);
}
exports.getPersistentValue = getPersistentValue;
/**
 * GetAllPersistentValues retrieves all persistent values.
 */
function getAllPersistentValues(ctx) {
    if (!ctx) {
        return {};
    }
    return (0, kv_1.getAll)(ctx, consts_1.persistentType);
}
exports.getAllPersistentValues = getAllPersistentValues;
/**
 * WithPersistentValue sets the value info the context by the given key.
 * This value will be propagated to the services along the RPC call chain.
 */
function withPersistentValue(ctx, k, v) {
    if (k.length === 0 || v.length === 0) {
        return ctx;
    }
    return (0, kv_1.addKV)(ctx, consts_1.persistentType, k, v);
}
exports.withPersistentValue = withPersistentValue;
/**
 * DelPersistentValue deletes a persistent key/value from the current context.
 * Since empty string value is not valid, we could just set the value to be empty.
 */
function delPersistentValue(ctx, k) {
    if (k.length === 0) {
        return ctx;
    }
    return (0, kv_1.addKV)(ctx, consts_1.persistentType, k, '');
}
exports.delPersistentValue = delPersistentValue;
