// @byted-service/metainfo/lib/kv.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.copyLink = exports.getAll = exports.getV = exports.addKV = exports.getKV = void 0;
const tslib_1 = require("tslib");
const deprecated_1 = require("./deprecated");
const ctx_converter_1 = require("./ctx_converter");
const helper = (0, tslib_1.__importStar)(require("./helper"));
function getKV(ctx) {
    if (ctx) {
        ctx_converter_1.ops.init(ctx);
        /** 兼容性代码 */
        if (helper.doubleWrite) {
            ctx = helper.convertMetainfoToNew(ctx);
        }
        /** 兼容性代码 */
        const pair = ctx_converter_1.ops.getCtxKey(ctx);
        if (pair) {
            return pair;
        }
    }
    return null;
}
exports.getKV = getKV;
function addKV(ctx, infoType, key, val) {
    if (!ctx) {
        return ctx;
    }
    ctx_converter_1.ops.setCtxKey(ctx, {
        pre: getKV(ctx),
        mode: infoType,
        key,
        val,
    });
    /** 兼容性代码 */
    if (helper.doubleWrite) {
        helper.addKVToOld(ctx, infoType, key, val);
    }
    /** 兼容性代码 */
    return ctx;
}
exports.addKV = addKV;
function getV(ctx, mode, key) {
    let kv = getKV(ctx);
    while (kv !== null) {
        if (kv.key === key && (kv.mode & mode) > 0) {
            return kv.val;
        }
        kv = kv.pre;
    }
    return '';
}
exports.getV = getV;
function getAll(ctx, mode) {
    const kvs = {};
    const del = {};
    let kv = getKV(ctx);
    while (kv) {
        if ((kv.mode & mode) > 0) {
            if (kvs[kv.key] === undefined) {
                kvs[kv.key] = kv.val;
                if (kv.val.length === 0) {
                    del[kv.key] = true;
                }
            }
        }
        kv = kv.pre;
    }
    const ks = Object.keys(del);
    for (let i = 0; i < ks.length; i++) {
        delete kvs[ks[i]];
    }
    return kvs;
}
exports.getAll = getAll;
/**
 * copyLink creates a new link copying the original one with each node processed by the given `modifier`.
 * If modifier returns nil, the kv pair will be discard.
 * Modifier must always return a valid pair object's pointer or nil.
 */
function copyLink(ctx, modifier) {
    let pre = null;
    let kv = getKV(ctx);
    while (kv) {
        const p = modifier(kv.mode, kv.key, kv.val);
        if (p !== null) {
            p.pre = pre;
            pre = p;
        }
        kv = kv.pre;
    }
    // When pre is nil after all, the whole link should be invisible in the new context
    ctx_converter_1.ops.setCtxKey(ctx, pre);
    /** 兼容性代码 */
    if (helper.doubleWrite) {
        ctx[deprecated_1.ctxKeyOld] = [];
        ctx[deprecated_1.UPSTREAM_META_INFO] = {};
        const copy = {};
        let kv = ctx_converter_1.ops.getCtxKey(ctx);
        while (kv) {
            const key = `${kv.key}-${kv.mode}`;
            if (copy[key] !== true) {
                helper.addKVToOld(ctx, kv.mode, kv.key, kv.val);
                copy[key] = true;
            }
            kv = kv.pre;
        }
    }
    /** 兼容性代码 */
    return ctx;
}
exports.copyLink = copyLink;
