// @universe-design/react/es/details-display/DetailsDisplay.js
/** @jsxRuntime classic */
/** @jsx jsx */
import * as React from 'react';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx } from '@universe-design/stylex/react';
import { DetailsDisplayContextProvider, useDetailsDisplay } from './DetailsDisplayContext';
import { useDetailsDisplayStylexStyles } from './sx';
import Row from './Row';

/**
 * @docgen
 */

/**
 * @docgen
 */

export const DetailsDisplay = props => {
  const {
    style,
    className,
    sx = [],
    ...useDetailsDisplayProp
  } = props;
  const contextValue = useDetailsDisplay(useDetailsDisplayProp);
  const {
    prefix,
    column,
    data
  } = contextValue;

  /**
   * format
   * 1、把每个数据的 span 格式化为 1 或者最大值 column；
   * 2、二维数组的每一个一维数组为一个tr；
   * 3、每一个数据的 span 大于 column 或者在每行的剩余空间放不下时，自动放在下一行；
   * 3.1、放到下一行后，会把上一行的最后一个数据扩大为占满剩余 tr 空间 span；
   * 4、最后输出二维数组；
   */

  const rowItems = ((originData, _column) => {
    const rows = [];
    let tmpRow = [];
    let rowRestCol = _column;
    originData.forEach((item, index) => {
      const span = Math.min(item.span || 1, _column);
      item.span = span;
      if (span <= rowRestCol) {
        tmpRow.push(item);
        rowRestCol -= span;
      } else {
        rows.push(tmpRow);
        tmpRow = [item];
        rowRestCol = _column - span;
      }
      if (rowRestCol === 0 || index === originData.length - 1) {
        rows.push(tmpRow);
        tmpRow = [];
        rowRestCol = _column;
      }
    });
    rows.forEach(item => {
      const sum = item.reduce((acc, child) => acc + (child.span || 1), 0);
      if (sum < _column) {
        item[item.length - 1].span += _column - sum;
      }
    });
    return rows;
  })(data, column);
  const styles = useDetailsDisplayStylexStyles({
    prefix
  });
  return jsx(DetailsDisplayContextProvider, {
    value: contextValue
  }, jsx("div", {
    className: className,
    sx: [styles.base, ...sx],
    style: style
  }, jsx("table", {
    sx: [styles.table]
  }, jsx("tbody", null, rowItems.map((row, r) => jsx(Row, {
    key: r,
    isRowEnd: r === rowItems.length - 1,
    row: row
  }))))));
};
