// @universe-design/react/es/details-display/Label.js
import React from "react";
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx } from '@universe-design/stylex/react';
import clsx from 'clsx';
import { useLocale } from '../config-provider/hooks';
import { useDetailsDisplayContext } from './DetailsDisplayContext';
import { useDetailsDisplayItemContext } from './DetailsDisplayItemContext';
import { useLabelStylexStyles } from './sx';
import { Text } from '../text';
const Label = () => {
  const locale = useLocale();
  const {
    prefix,
    colon,
    labelStyle: rootLabelStyle,
    labelClassName: rootLabelClassName
  } = useDetailsDisplayContext();
  const {
    label,
    labelStyle,
    labelClassName
  } = useDetailsDisplayItemContext();
  const styles = useLabelStylexStyles({
    prefix: `${prefix}__label`
  });
  return jsx("div", {
    sx: [styles.base]
  }, jsx(Text, {
    style: {
      ...rootLabelStyle,
      ...labelStyle
    },
    className: clsx(rootLabelClassName, labelClassName),
    lineClamp: 2
  }, label), jsx(Text, {
    sx: [styles.colon]
  }, colon && locale.global.colon));
};
export default Label;
