// @universe-design/react/es/details-display/Row.js
/* eslint-disable react/no-unknown-property */
/** @jsxRuntime classic */
/** @jsx jsx */
/** @jsxFrag React.Fragment */
import * as React from 'react';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx } from '@universe-design/stylex/react';
import { useDetailsDisplayContext } from './DetailsDisplayContext';
import { DetailsDisplayItemContextProvider, useDetailsDisplayItem } from './DetailsDisplayItemContext';
import { useRowStylexStyles } from './sx';
import HeaderCell from './HeaderCell';
import DataCell from './DataCell';
import Label from './Label';
import Content from './Content';
const renderCells = (items, _ref) => {
  let {
    type,
    component,
    showLabel,
    showContent,
    isRowEnd
  } = _ref;
  return items.map((item, i) => {
    const {
      layout
    } = useDetailsDisplayContext();
    const layoutInline = layout === 'inline';
    // 处理cell之间的间隙，是否是最后一列
    const isColumnEnd = i === items.length - 1;
    const contextValue = useDetailsDisplayItem({
      ...item,
      isRowEnd,
      isColumnEnd
    });
    if (typeof component === 'string') {
      return jsx(DetailsDisplayItemContextProvider, {
        key: `${type}-${component}-${i}`,
        value: contextValue
      }, showLabel && !layoutInline && jsx(HeaderCell, null, jsx(Label, null)), showContent && jsx(DataCell, null, layoutInline && jsx(Label, null), jsx(Content, null)));
    }
    return [jsx(DetailsDisplayItemContextProvider, {
      key: `${type}-${component[0]}-${i}`,
      value: {
        ...contextValue,
        span: 1
      }
    }, jsx(HeaderCell, null, jsx(Label, null))), jsx(DetailsDisplayItemContextProvider, {
      key: `${type}-${component[1]}-${i}`,
      value: {
        ...contextValue,
        span: (item.span || 1) * 2 - 1
      }
    }, jsx(DataCell, null, layoutInline && jsx(Label, null), jsx(Content, null)))];
  });
};
const Row = props => {
  const {
    isRowEnd,
    row
  } = props;
  const {
    prefix,
    layout
  } = useDetailsDisplayContext();
  const styles = useRowStylexStyles({
    prefix: `${prefix}__row`
  });
  if (layout === 'vertical') {
    return jsx(React.Fragment, null, jsx("tr", {
      sx: [styles.base]
    }, renderCells(row, {
      type: 'label',
      component: 'th',
      showLabel: true,
      isRowEnd: false
    })), jsx("tr", {
      sx: [styles.base]
    }, renderCells(row, {
      type: 'content',
      component: 'td',
      showContent: true,
      isRowEnd
    })));
  }
  if (layout === 'inline') {
    return jsx("tr", {
      sx: [styles.base]
    }, renderCells(row, {
      type: 'inline',
      component: 'td',
      showLabel: true,
      showContent: true,
      isRowEnd
    }));
  }
  return jsx("tr", {
    sx: [styles.base]
  }, renderCells(row, {
    type: 'item',
    component: ['th', 'td'],
    isRowEnd
  }));
};
export default Row;
