// @universe-design/react/es/details-display/sx.js
//.ujw5fx7{width:16px;}
import ut0j16j from "./../.stylex/ut0j16j";
//.uq2vo04{min-width:16px;}
import u5q3evl from "./../.stylex/u5q3evl";
//.u1rvp7gu{color:@text-caption;}
import ueh72r3 from "./../.stylex/ueh72r3";
//.utuwxu6{display:inline-flex;}
import ug0oh3b from "./../.stylex/ug0oh3b";
//.u1guyajx{align-items:baseline;}
import u1gmpn6f from "./../.stylex/u1gmpn6f";
//.unaqv0d{padding-right:40px;}
import u4vqhs6 from "./../.stylex/u4vqhs6";
//.u15kemlc{padding-right:24px;}
import u127hhy3 from "./../.stylex/u127hhy3";
//.u1y2xyjm{padding-bottom:4px;}
import u1ibfn25 from "./../.stylex/u1ibfn25";
//.ues3tcz{text-align:right;}
import u12v4awa from "./../.stylex/u12v4awa";
//.u1o700av{text-align:left;}
import u135sb2c from "./../.stylex/u135sb2c";
//.ue2my4m{padding-bottom:20px;}
import u1co87pz from "./../.stylex/u1co87pz";
//.up67ikv{padding-bottom:16px;}
import u1lnu8wd from "./../.stylex/u1lnu8wd";
//.upe6lb7{padding-bottom:12px;}
import u1hdaclu from "./../.stylex/u1hdaclu";
//.u1z30l8{color:@text-title;}
import u22yt5q from "./../.stylex/u22yt5q";
//.u1nbblvp{padding-top:0px;}
import u165peiw from "./../.stylex/u165peiw";
//.uifp7yv{padding-right:0px;}
import u24on9z from "./../.stylex/u24on9z";
//.u1ov4xf1{padding-bottom:0px;}
import uu1oa8u from "./../.stylex/uu1oa8u";
//.u1asdtw4{padding-left:0px;}
import u1pwi6nc from "./../.stylex/u1pwi6nc";
//.uvrlu0f{min-width:44px;}
import u11kywcj from "./../.stylex/u11kywcj";
//.u82itaf{line-height:22px;}
import u1dhzo8t from "./../.stylex/u1dhzo8t";
//.u1cdqgvm{vertical-align:baseline;}
import u1iw9z84 from "./../.stylex/u1iw9z84";
//.uly5x3f{width:100%;}
import u13d3rh8 from "./../.stylex/u13d3rh8";
//.uses1vf{font-size:14px;}
import u19bueuo from "./../.stylex/u19bueuo";
import stylex from '@universe-design/stylex';
import { makeStylexStyles } from '@universe-design/stylex/react';

// table
export const useDetailsDisplayStylexStyles = makeStylexStyles(_ref => {
  let {
    prefix
  } = _ref;
  return {
    base: {
      "props": {
        "Be2twd7": u19bueuo
      },
      "alias": prefix
    },
    table: {
      "props": {
        "a9b677": u13d3rh8
      },
      "alias": `${prefix}__table`
    }
  };
});

// row
export const useRowStylexStyles = makeStylexStyles(_ref2 => {
  let {
    prefix
  } = _ref2;
  return {
    base: {
      "props": {
        "Bg96gwp": u1dhzo8t,
        "ha4doy": u1iw9z84
      },
      "alias": prefix
    }
  };
});

// HeaderCell && DataCell
export const useCellStylexStyles = makeStylexStyles(_ref3 => {
  let {
    prefix
  } = _ref3;
  return {
    /**
     * th & reset
     * minWidth：两个汉字（14 * 2） + 冒号的宽度（16）
     *  */
    th: {
      "props": {
        "z8tnut": u165peiw,
        "z189sj": u24on9z,
        "Byoj8tv": uu1oa8u,
        "uwmqm3": u1pwi6nc,
        "Bf4jedk": u11kywcj
      },
      "alias": `${prefix}`
    },
    // td & reset
    td: {
      "props": {
        "z8tnut": u165peiw,
        "z189sj": u24on9z,
        "Byoj8tv": uu1oa8u,
        "uwmqm3": u1pwi6nc,
        "sj55zd": u22yt5q,
        "Bf4jedk": u11kywcj
      },
      "alias": `${prefix}`
    },
    // 列表三种尺寸间距
    size: {
      small: {
        "props": {
          "Byoj8tv": u1hdaclu
        },
        "alias": `${prefix}-small`
      },
      medium: {
        "props": {
          "Byoj8tv": u1lnu8wd
        },
        "alias": `${prefix}-medium`
      },
      large: {
        "props": {
          "Byoj8tv": u1co87pz
        },
        "alias": `${prefix}-large`
      }
    },
    align: {
      left: {
        "props": {
          "fsow6f": u135sb2c
        },
        "alias": `${prefix}-align-left`
      },
      right: {
        "props": {
          "fsow6f": u12v4awa
        },
        "alias": `${prefix}-align-right`
      }
    },
    // 垂直布局 label 和 content 的间距
    headerGap: {
      "props": {
        "Byoj8tv": u1ibfn25
      },
      "alias": `${prefix}-gap`
    },
    layout: {
      // 垂直布局横向 item 间距
      verticalGap: {
        "props": {
          "z189sj": u127hhy3
        },
        "alias": `${prefix}-vertical-gap`
      },
      // 水平布局横向 item 间距
      horizontalGap: {
        "props": {
          "z189sj": u4vqhs6
        },
        "alias": `${prefix}-horizontal-gap`
      }
    },
    // layout === 'inline'时，设置display: 'inline-flex'
    inlineFlex: {
      "props": {
        "mc9l5x": ug0oh3b,
        "Bt984gj": u1gmpn6f
      },
      "alias": `${prefix}-inline-flex`
    }
  };
});

// label
export const useLabelStylexStyles = makeStylexStyles(_ref4 => {
  let {
    prefix
  } = _ref4;
  return {
    base: {
      "props": {
        "mc9l5x": ug0oh3b,
        "Bt984gj": u1gmpn6f,
        "sj55zd": ueh72r3
      },
      "alias": prefix
    },
    colon: {
      "props": {
        "a9b677": ut0j16j,
        "Bf4jedk": u5q3evl,
        "fsow6f": u135sb2c
      },
      "alias": `${prefix}__colon`
    }
  };
});
