// @universe-design/react/src/tree/stories/homeless.stories.tsx
import type { Meta } from '@storybook/react';
import type { TreeProps, TreeRef } from '@universe-design/react';
import { Button, Tree } from '@universe-design/react';
import * as React from 'react';

export const RepeatKeys = () => {
  const treeData: TreeProps['treeData'] = [
    {
      label: 'Parent 1',
      value: '0-0',
      children: [
        {
          label: 'Parent 1-1',
          value: '0-0',
        },
        {
          label: 'Parent 1-2',
          value: '0-0',
        },
      ],
    },
    {
      label: 'Parent 2',
      value: 'parent-2',
    },
  ];

  return (
    <Tree
      searchValue="1"
      onSelect={keys => console.log(keys)}
      onCheck={keys => console.log(keys)}
      treeData={treeData}
    />
  );
};

const getChildren = (prefix: string) =>
  new Array(20).fill('').map((_, index) => ({
    label: `${prefix} child ${index}`,
    value: `${prefix}_${index}`,
  }));
export const ScrollInfoView = () => {
  const ref = React.useRef<TreeRef>(null);
  const treeData: TreeProps['treeData'] = [
    {
      label: 'Parent 1',
      value: '0-1',
      children: getChildren('1'),
    },
    {
      label: 'Parent 2',
      value: '0-2',
      children: getChildren('2'),
    },
  ];

  return (
    <>
      <div>
        <Button onClick={() => ref.current?.scrollIntoView({ type: 'node', key: '1_10' })}> ScrollTo: 1-10</Button>
        <Button onClick={() => ref.current?.scrollIntoView({ type: 'node', key: '2_10' })}> ScrollTo: 2-10</Button>
        <Button onClick={() => ref.current?.scrollIntoView(10)}> index: 10</Button>
      </div>
      <Tree
        height={200}
        ref={ref}
        searchValue="1"
        onSelect={keys => console.log(keys)}
        onCheck={keys => console.log(keys)}
        treeData={treeData}
      />
    </>
  );
};
export default {
  title: 'Components/Tree/临时示例',
  component: Tree,
} satisfies Meta;
