// @universe-design/react/src/tree/stories/screenshot.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C9Disabled from '../demo/9.disabled';
import CLongTextScreenshot from '../demo/long-text.screenshot';
import CTreeRootHasNoNonLeafChildren from '../demo/tree-root-has-no-non-leaf-children';
import CZCalendarMeetingRoomSelect from '../demo/z.calendar-meeting-room-select';
if (!(C9Disabled as Story).storyName) {
  (C9Disabled as Story).storyName = '9 disabled';
}
if (!(CLongTextScreenshot as Story).storyName) {
  (CLongTextScreenshot as Story).storyName = 'Long text screenshot';
}
if (!(CTreeRootHasNoNonLeafChildren as Story).storyName) {
  (CTreeRootHasNoNonLeafChildren as Story).storyName = 'Tree root has no non leaf children';
}
if (!(CZCalendarMeetingRoomSelect as Story).storyName) {
  (CZCalendarMeetingRoomSelect as Story).storyName = 'Z calendar meeting room select';
}

export default {
  title: 'Components/Tree/截图测试',
  tags: ['autodocs'],
};

export { C9Disabled, CLongTextScreenshot, CTreeRootHasNoNonLeafChildren, CZCalendarMeetingRoomSelect };
