// @universe-design/react/src/tree/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Basic from '../demo/0.basic';
import C1Controlled from '../demo/1.controlled';
import C11Search from '../demo/11.search';
import C2CheckStrictly from '../demo/2.check-strictly';
import C3Dynamic from '../demo/3.dynamic';
import C8SwitcherIcon from '../demo/8.switcher-icon';
import CDisabledSelectParent from '../demo/disabled-select-parent';
import CNodeRenderer from '../demo/node-renderer';
if (!(C0Basic as Story).storyName) {
  (C0Basic as Story).storyName = '0. 基础用法';
  (C0Basic as Story).parameters = {
    docs: {
      description: {
        story: '简单的用法。展示可勾选、选中、禁用、默认展开等功能。\n\n',
      },
    },
  };
}
if (!(C1Controlled as Story).storyName) {
  (C1Controlled as Story).storyName = '1. 受控操作';
  (C1Controlled as Story).parameters = {
    docs: {
      description: {
        story: '受控操作示例。\n\n',
      },
    },
  };
}
if (!(C11Search as Story).storyName) {
  (C11Search as Story).storyName = '11. 搜索树';
  (C11Search as Story).parameters = {
    docs: {
      description: {
        story: '树组件支持搜索和高亮。\n\n',
      },
    },
  };
}
if (!(C2CheckStrictly as Story).storyName) {
  (C2CheckStrictly as Story).storyName = '2. 严格勾选模式';
  (C2CheckStrictly as Story).parameters = {
    docs: {
      description: {
        story: '父子元素之间没有关联。\n\n',
      },
    },
  };
}
if (!(C3Dynamic as Story).storyName) {
  (C3Dynamic as Story).storyName = '3. 异步数据加载';
  (C3Dynamic as Story).parameters = {
    docs: {
      description: {
        story: '点击展开节点，动态加载数据。\n\n',
      },
    },
  };
}
if (!(C8SwitcherIcon as Story).storyName) {
  (C8SwitcherIcon as Story).storyName = '8. 自定义展开 / 折叠图标';
  (C8SwitcherIcon as Story).parameters = {
    docs: {
      description: {
        story: '自定义展开 / 折叠图标。\n\n',
      },
    },
  };
}
if (!(CDisabledSelectParent as Story).storyName) {
  (CDisabledSelectParent as Story).storyName = '10. 通过组合的方式禁止选中父节点';
  (CDisabledSelectParent as Story).parameters = {
    docs: {
      description: {
        story: '可以通过组合的方式做到禁止选中父节点但是保留正常的样式\n\n',
      },
    },
  };
}
if (!(CNodeRenderer as Story).storyName) {
  (CNodeRenderer as Story).storyName = '9. 完全组合树节点';
  (CNodeRenderer as Story).parameters = {
    docs: {
      description: {
        story: '当默认节点布局不适用于你的场景时，可以尝试用子组件组合的方式去编写。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Tree/官网示例',
  tags: ['autodocs'],
};

export {
  C0Basic,
  C1Controlled,
  C11Search,
  C2CheckStrictly,
  C3Dynamic,
  C8SwitcherIcon,
  CDisabledSelectParent,
  CNodeRenderer,
};
