// @byted-service/bytedtrace/lib/breaker/breakConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BreakerConfigurator = void 0;
const tslib_1 = require("tslib");
const confProvider = tslib_1.__importStar(require("../confProvider"));
class BreakerConfigurator {
    constructor(configurator) {
        this.configurator = configurator;
    }
    async syncConfig() {
        const cf = await this.configurator.pullBreakerConfig(true);
        this.init(cf);
    }
    asyncConfig() {
        this.init(confProvider.DEFAULT_CB_CONFIG);
    }
    init(cf) {
        var _a;
        this.maxSpanCountInTxn = cf.maxSpanCountInTxn;
        this.maxTagKVCountInObj = cf.maxTagKVCountInObj;
        this.maxEventCountInSpan = cf.maxEventCountInSpan;
        this.maxCustomMetricCountInSpan = cf.maxCustomMetricCountInSpan;
        this.maxTagValueLen = cf.maxTagValueLen;
        this.maxEventDataLen = cf.maxEventDataLen;
        this.maxTxnDataLen = cf.maxTxnDataLen;
        this.maxTxnSamplingRatePerOperation = cf.maxTxnSamplingRatePerOperation;
        this.maxPostTraceSamplingRatePerOperation =
            cf.maxPostTraceSamplingRatePerOperation;
        this.maxDyeingSamplingRatePerOperation =
            cf.maxDyeingSamplingRatePerOperation;
        const limiterConfigs = [];
        (_a = cf.maxPostTraceSamplingDetailStrategies) === null || _a === void 0 ? void 0 : _a.forEach((cfs) => {
            limiterConfigs.push({
                spanType: cfs.spanType,
                spanName: cfs.spanName,
                hitPerMinute: cfs.thresholdPerMin,
            });
        });
        this.postTraceLimitConfigs = limiterConfigs;
    }
}
exports.BreakerConfigurator = BreakerConfigurator;
