// @byted-service/bytedtrace/lib/breaker/breaker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Breaker = void 0;
const tslib_1 = require("tslib");
const bytedtracePb = tslib_1.__importStar(require("../gen/bytedtrace_pb"));
const sampler = tslib_1.__importStar(require("../sampler"));
const utils = tslib_1.__importStar(require("../utils"));
const breakConfig_1 = require("./breakConfig");
class Breaker {
    constructor(env, tracer, logger) {
        this.env = env;
        this.logger = logger;
        this.tracer = tracer;
    }
    /**
     * 先从 tcc 拉取配置，再初始化 limiter
     * @param configurator
     */
    async syncInit(configurator) {
        this.config = new breakConfig_1.BreakerConfigurator(configurator);
        await this.config.syncConfig();
        this.initLimiterPack();
    }
    /**
     * 先使用默认配置再拉取 tcc 配置
     * @param configurator
     */
    asyncInit(configurator) {
        this.config = new breakConfig_1.BreakerConfigurator(configurator);
        this.config.asyncConfig();
        this.initLimiterPack();
        this.syncConfig({});
    }
    async syncConfig(_) {
        await this.config.syncConfig();
        this.initLimiterPack();
    }
    getMaxSpanCountInTxn() {
        return this.config.maxSpanCountInTxn;
    }
    getMaxTagKVCountInObj() {
        return this.config.maxTagKVCountInObj;
    }
    getMaxEventCountInSpan() {
        return this.config.maxEventCountInSpan;
    }
    getMaxCustomMetricCountInSpan() {
        return this.config.maxCustomMetricCountInSpan;
    }
    truncatedStringToBytesContent(data) {
        data = data.slice(0, this.config.maxEventDataLen);
        return Buffer.from(data);
    }
    buildTags(tagKVs) {
        if (tagKVs.length <= 0) {
            return [];
        }
        const tagPbs = [];
        for (let i = 0; i < tagKVs.length; i++) {
            const [key, val] = tagKVs[i];
            tagPbs.push(this.buildTag(key, val));
        }
        return tagPbs;
    }
    buildTag(key, value) {
        const tag = new bytedtracePb.Tag();
        tag.setKey(key);
        let type = typeof value;
        if (type === 'object' && value instanceof Buffer) {
            type = 'bytes';
        }
        switch (type) {
            case 'string':
                tag.setVStr(this.truncatedStringTag(value));
                tag.setTagType(bytedtracePb.TagType.STRING);
                break;
            case 'bytes':
                tag.setVBytes(this.truncatedBytesTag(value));
                tag.setTagType(bytedtracePb.TagType.BYTES);
                break;
            case 'number':
                tag.setVDouble(value);
                tag.setTagType(bytedtracePb.TagType.DOUBLE);
                break;
            case 'boolean':
                tag.setVBool(value);
                tag.setTagType(bytedtracePb.TagType.BOOL);
                break;
            default:
                tag.setVStr(this.truncatedStringTag(utils.stringify(value)));
                tag.setTagType(bytedtracePb.TagType.STRING);
        }
        return tag;
    }
    initLimiterPack() {
        const postTraceLimiterPack = new sampler.LimiterPack(this.tracer.options.maxOperationForLimiter, sampler.DEFAULT_EXPIRED_TIME_MS, sampler.newAtomicLimiterFunc(this.config.maxPostTraceSamplingRatePerOperation, -1));
        // PostTrace 细分条件设置
        this.config.postTraceLimitConfigs.forEach((cfs) => {
            postTraceLimiterPack.insertLimiter(`${cfs.spanType}:${cfs.spanName}`, new sampler.TickLimiter(sampler.newAtomicLimiterFunc(cfs.hitPerMinute / 60.0, -1), -1));
        });
        const debugTraceLimiterPack = new sampler.LimiterPack(this.tracer.options.maxOperationForLimiter, sampler.DEFAULT_EXPIRED_TIME_MS, sampler.newAtomicLimiterFunc(this.config.maxDyeingSamplingRatePerOperation, -1));
        const txnSamplingLimiterPack = new sampler.LimiterPack(this.tracer.options.maxOperationForLimiter, sampler.DEFAULT_EXPIRED_TIME_MS, sampler.newAtomicLimiterFunc(this.config.maxTxnSamplingRatePerOperation, -1));
        this.postTraceLimiterPack = postTraceLimiterPack;
        this.debugTraceLimiterPack = debugTraceLimiterPack;
        this.txnSamplingLimiterPack = txnSamplingLimiterPack;
    }
    truncatedStringTag(value) {
        return value === null || value === void 0 ? void 0 : value.slice(0, this.config.maxTagValueLen);
    }
    truncatedBytesTag(value) {
        return value === null || value === void 0 ? void 0 : value.slice(0, this.config.maxTagValueLen);
    }
}
exports.Breaker = Breaker;
