// @universe-design/react/src/card/index.tsx
import type { CSSProperties, HTMLAttributes } from 'react';
import React, { forwardRef } from 'react';
import clsx from 'clsx';
import { useCssScope } from '../config-provider/hooks';
import type { DefaultSizeType } from '../util';
import { getDefaultSizeAbbr } from '../util';

import './style';

/**
 * @public
 */
export interface CardProps extends HTMLAttributes<HTMLDivElement> {
  /**
   * 节点样式
   */
  style?: CSSProperties;
  /**
   * 节点类名
   */
  className?: string;
  /**
   * 是否渲染阴影
   * @defaultValue true
   */
  shadowed?: boolean;
  /**
   * 是否渲染边框
   * @defaultValue true
   */
  bordered?: boolean;
  /**
   * 是否关闭鼠标 hover 时的交互状态
   * @defaultValue false
   */
  noHoverState?: boolean;
  /**
   * 阴影大小
   * @defaultValue `medium`
   */
  shadow?: DefaultSizeType;
  /**
   * 圆角大小
   * @defaultValue `small`
   */
  borderRadius?: DefaultSizeType;
}

export const Card = forwardRef<HTMLDivElement, CardProps>((props, ref) => {
  const {
    className,
    shadowed = true,
    bordered = true,
    shadow = 'medium',
    borderRadius = 'small',
    noHoverState = false,
    ...restProps
  } = props;

  const cssScope = useCssScope('card');
  const usedShadow = getDefaultSizeAbbr(shadow, 'medium');
  const usedBorderRadius = getDefaultSizeAbbr(borderRadius, 'small');

  const cls = clsx(
    cssScope(),
    {
      [cssScope(`--shadow-${usedShadow}`)]: usedShadow,
      [cssScope(`--br-${usedBorderRadius}`)]: usedBorderRadius,
      [cssScope('--noborder')]: !bordered,
      [cssScope('--noshadows')]: !shadowed,
      [cssScope('--no-hover-state')]: noHoverState,
    },
    className,
  );

  return <div ref={ref} className={cls} {...restProps} />;
});
