// @universe-design/react/src/steps/NavigationIndicator.tsx
import stylex from '@universe-design/stylex';
import * as React from 'react';
import { useStepsContext } from './useSteps';
const IndicatorColor = `rgba(${$token['N900-raw']}, 15%)`;
const IndicatorActiveColor = $token['primary-content-default'];

const sxIndicator = stylex.create({
  position: 'relative',
  backgroundColor: IndicatorColor,
  height: 1,
  marginTop: 9,
});

const sxIndicatorActive = stylex.create({
  ':before': {
    content: "''",
    position: 'absolute',
    bottom: 0,
    left: 24,
    right: 0,
    display: 'inline-block',
    height: 3,
    backgroundColor: IndicatorActiveColor,
    transition: 'width .3s, left .3s',
    transitionTimingFunction: 'ease-out',
  },
});

export default (props: { index: number }) => {
  const { index } = props;
  const { activeIndex, showNavIndicator } = useStepsContext();
  if (!showNavIndicator) {
    return null;
  }
  return (
    <div
      data-testid="indicator"
      className={stylex(
        sxIndicator,
        activeIndex === index && sxIndicatorActive,
        index === 0 && stylex.create({ ':before': { left: 0 } }),
      )}
    />
  );
};
