// @universe-design/react/src/steps/step.tsx
/** @jsxRuntime classic */
/** @jsx jsx */
/* eslint-disable max-lines-per-function */
import type { ReactNode } from 'react';
import React, { forwardRef } from 'react';
import clsx from 'clsx';
import DoneOutlined from '@universe-design/icons-react/DoneOutlined';
import CloseOutlined from '@universe-design/icons-react/CloseOutlined';
import RightBoldOutlined from '@universe-design/icons-react/RightBoldOutlined';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx } from '@universe-design/stylex/react';
import type { StepStatus } from './interface';
import stylex from '@universe-design/stylex';
import { cssTypographyHeadline, cssTypographyBody0 } from '../stylex';
import { useStepsContext } from './useSteps';
import Description from './Description';
import { Text } from '../text';

/**
 * `<Step />` 组件的属性。
 * @public
 */
export interface StepProps {
  /**
   * `className` 属性透传
   */
  className?: string;
  /**
   * 步骤图标
   */
  icon?: ReactNode;
  /**
   * 标题
   */
  title: ReactNode;
  /**
   * 子标题
   */
  subTitle?: ReactNode;
  /**
   * 描述
   */
  description?: ReactNode;
  /**
   * 当前节点的编号
   */
  stepIndex?: number;
  prefixCls?: string;
  /**
   * 状态。当不配置该属性时，会使用父级 `Steps` 的 `status` 作为 `status` 值。
   * @defaultValue `wait`
   */
  status?: StepStatus;
  /**
   * 禁用点击
   * @defaultValue false
   */
  disabled?: boolean;
}

const Step = forwardRef<HTMLDivElement, StepProps>((props, ref) => {
  const { className, icon, title, subTitle, description, stepIndex = 0, status = 'wait', disabled } = props;
  const {
    registerNode,
    separateDescription,
    fixedDirection,
    clickable,
    steps,
    cssScope,
    activeIndex,
    showIcon,
    showNavIndicator,
    type,
    count,
    initial,
    progressDot,
    onStepClick,
  } = useStepsContext();
  const prefixCls = cssScope();
  const index = stepIndex - initial;

  React.useEffect(() => {
    registerNode(stepIndex, {
      description,
      disabled,
      status,
    });
  }, [description, disabled, registerNode, stepIndex, status]);

  const isLast = count - 1 === index;

  const preStatus = steps[index - 1]?.status;

  const wrapperCls = clsx(
    `${prefixCls}__item`,
    `${prefixCls}__item--${status}`,
    className,
    {
      [`${prefixCls}__item--last`]: isLast,
      [`${prefixCls}--pre-${preStatus}`]: preStatus,
      [`${prefixCls}__item--disabled`]: disabled,
      [`${prefixCls}__item--active`]: stepIndex === activeIndex,
      [`${prefixCls}__item--with-indicator`]: type === 'navigation' && showNavIndicator,
      [`${prefixCls}__item--without-icon`]: type === 'navigation' && !showIcon,
      [`${prefixCls}__item--clickable`]: clickable && !disabled,
    },
    ['process', 'error'].includes(status) ? stylex(cssTypographyHeadline) : stylex(cssTypographyBody0),
  );

  function handleClick() {
    onStepClick(stepIndex);
  }

  function renderIconNode() {
    let iconNode: ReactNode = stepIndex + 1;
    if (type === 'dot') {
      iconNode = <span className={`${prefixCls}__icon-dot`} />;
    } else if (icon) {
      iconNode = icon;
    } else if (status === 'finish') {
      iconNode = <DoneOutlined />;
    } else if (status === 'error') {
      iconNode = <CloseOutlined />;
    }
    const cls = clsx(`${prefixCls}__icon`, {
      [`${prefixCls}--include-icon`]: typeof iconNode !== 'number',
    });
    return <span className={cls}>{iconNode}</span>;
  }

  function renderIcon() {
    let iconWrapper: ReactNode;
    if (type === 'navigation' && !showIcon) {
      return null;
    }
    iconWrapper = <div className={`${prefixCls}__item-icon`}>{renderIconNode()}</div>;
    if (typeof progressDot === 'function') {
      iconWrapper = progressDot(iconWrapper, {
        index: stepIndex,
        status,
        title,
        subTitle,
        description,
      });
    }
    return iconWrapper;
  }

  return (
    <div
      className={wrapperCls}
      ref={ref}
      sx={[
        separateDescription &&
          stylex.create({
            display: 'flex',
            alignItems: 'center',
            height: '100%',
          }),
      ]}
    >
      <div onClick={handleClick} className={`${prefixCls}__item-container`}>
        <div className={`${prefixCls}__item-content`}>
          {!isLast && <div className={`${prefixCls}__item-tail`} />}
          {renderIcon()}
          <div className={`${prefixCls}__item-title`}>
            <Text
              className={`${prefixCls}__item-title-inner`}
              as="div"
              lineClamp={fixedDirection === 'horizontal' ? 2 : undefined}
              variant="plain"
            >
              {title}
              {subTitle && <span className={`${prefixCls}__item-subtitle`}>{subTitle}</span>}
            </Text>
          </div>
        </div>
        {!separateDescription && description && <Description stepIndex={stepIndex} />}
        {!separateDescription && !isLast && <div className={`${prefixCls}__item-padding`} />}
      </div>
      {type === 'navigation' && !isLast && (
        <div className={`${prefixCls}__item-arrow`}>
          <RightBoldOutlined />
        </div>
      )}
    </div>
  );
});

export default Step;
