// @universe-design/styles/ts/vc-livestream/index.light.atom.ts
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
export const vctoken_livestream_bg_addnew_pressed = '#eff0f1';
export const vctoken_livestream_bg_addscene_hover = '#f0f4ff';
export const vctoken_livestream_bg_addscene_pressed = '#f0f4ff';
export const vctoken_livestream_bg_applaud_emphasis = '#ffffff';
export const vctoken_livestream_bg_btn = '#f5f6f7';
export const vctoken_livestream_bg_btn_pressed = 'rgba(51, 109, 244, 0.2)';
export const vctoken_livestream_bg_btn_selected = '#e0e9ff';
export const vctoken_livestream_bg_cam_off = '#f2f3f5';
export const vctoken_livestream_bg_comment_fg_hover = 'rgba(239, 240, 241, 0)';
export const vctoken_livestream_bg_comment_hover = '#eff0f1';
export const vctoken_livestream_bg_content = '#cfd2d6';
export const vctoken_livestream_bg_core = '#e8e9eb';
export const vctoken_livestream_bg_core_speaker = 'rgba(187, 191, 196, 0.7)';
export const vctoken_livestream_bg_gradient_fg_emphasis = 'rgba(0, 0, 0, 0)';
export const vctoken_livestream_bg_input_activity = '#ffffff';
export const vctoken_livestream_bg_input_emphasis = 'rgba(31, 35, 41, 0.05)';
export const vctoken_livestream_bg_interaction = '#ffffff';
export const vctoken_livestream_bg_members = '#f5f6f7';
export const vctoken_livestream_bg_player_button_emphasis = 'rgba(31, 35, 41, 0.8)';
export const vctoken_livestream_bg_player_fill_emphasis = '#ebecee';
export const vctoken_livestream_bg_reaction_left = '#3370ff';
export const vctoken_livestream_bg_reaction_right = '#7edafb';
export const vctoken_livestream_bg_speaker = '#e0e9ff';
export const vctoken_livestream_bg_tag = 'rgba(51, 109, 244, 0.2)';
export const vctoken_livestream_bg_tag_livestreaming = '#fcd4d3';
export const vctoken_livestream_bg_tag_testing = '#cfd1f7';
export const vctoken_livestream_bg_theme_emphasis = 'rgba(255, 255, 255, 0.95)';
export const vctoken_livestream_bg_top_set = '#eff0f1';
export const vctoken_livestream_bg_top_set_1 = '#f5f6f7';
export const vctoken_livestream_bg_top_set_2 = '#f5f6f7';
export const vctoken_livestream_fill_btn1 = '#ffffff';
export const vctoken_livestream_fill_btn2 = '#eff0f1';
export const vctoken_livestream_fill_min_player_mask = 'rgba(0, 0, 0, 0.45)';
export const vctoken_livestream_fill_new_tips_hover_web = '#f2f3f5';
export const vctoken_livestream_fill_new_tips_normal_web = '#ffffff';
export const vctoken_livestream_fill_playbar_volume_tip_hover = 'rgba(76, 136, 255, 0.4)';
export const vctoken_livestream_fill_player_button_hover_1 = 'rgba(76, 136, 255, 0.2)';
export const vctoken_livestream_fill_player_button_hover_2 = 'rgba(235, 235, 235, 0.1)';
export const vctoken_livestream_fill_player_button_pressed_1 = 'rgba(76, 136, 255, 0.3)';
export const vctoken_livestream_fill_player_button_pressed_2 = 'rgba(235, 235, 235, 0.2)';
export const vctoken_livestream_fill_player_panel_activity = 'rgba(31, 35, 41, 0.9)';
export const vctoken_livestream_fill_player_tag_finish = '#f2962c';
export const vctoken_livestream_fill_player_tag_live = '#f05b56';
export const vctoken_livestream_fill_preview_color_1 = '#f2f3f5';
export const vctoken_livestream_fill_preview_color_2 = '#f2f3f5';
export const vctoken_livestream_fill_tab_color_1 = '#ffffff';
export const vctoken_livestream_fill_tab_color_2 = 'rgba(255, 255, 255, 0)';
export const vctoken_livestream_fill_top_set_border = '#dee0e3';
export const vctoken_livestream_icon_btn_selected = '#336df4';
export const vctoken_livestream_icon_btn_unselected = '#2b2f36';
export const vctoken_livestream_icon_more = '#ffffff';
export const vctoken_livestream_icon_more_hover = '#f5f6f7';
export const vctoken_livestream_icon_pressed = '#1456f0';
export const vctoken_livestream_line_btn_divider = '#c2d4ff';
export const vctoken_livestream_line_btn_selected = 'rgba(148, 180, 255, 0.7)';
export const vctoken_livestream_line_mainscreen = 'rgba(255, 255, 255, 0.4)';
export const vctoken_livestream_line_speaker = 'rgba(148, 180, 255, 0.7)';
export const vctoken_livestream_player_shadow_emphasis = '0px 0px 1px rgba(0, 0, 0, 0.5)';
export const vctoken_livestream_player_shadow_emphasis_1 =
  '0px 4px 8px rgba(31, 35, 41, 0.06), 0px 8px 24px 8px rgba(31, 35, 41, 0.04), 0px 6px 12px rgba(31, 35, 41, 0.04)';
export const vctoken_livestream_player_shadow_emphasis_2 = '0px 1px 4px rgba(0, 0, 0, 0.2)';
export const vctoken_livestream_player_text_shadow_1 = '0px 0px 1px rgba(0, 0, 0, 0.3)';
export const vctoken_livestream_preview_inner_shadow_default = '0px 4px 24px rgba(187, 191, 196, 0.18)';
export const vctoken_livestream_preview_shadow_ash_blue = '0px 8px 30px rgba(14, 24, 40, 0.26)';
export const vctoken_livestream_preview_shadow_default = '0px 8px 48px rgba(31, 35, 41, 0.12)';
export const vctoken_livestream_text_btn_selected = '#336df4';
export const vctoken_livestream_text_btn_unselected = '#1f2329';
export const vctoken_livestream_text_logo_audience = '#000000';
export const vctoken_livestream_text_playbar_selected = '#4c88ff';
export const vctoken_livestream_text_pressed = '#1456f0';
export const vctoken_livestream_text_reaction_number = '#6425d0';
export const vctoken_livestream_text_tag = '#1456f0';
export const vctoken_livestream_web_commemt_inset_shadow = 'inset -2px 2px 2px rgba(255, 255, 255, 0.9)';
export const vctoken_livestream_web_title_inset_shadow = 'inset 0px 2px 2px rgba(255, 255, 255, 0.9)';
