// @universe-design/react/src/image-viewer/constants.ts
import type { ImageViewerToolbarProps } from './useImageViewerToolbar';

// 缩放递增常量
export const SCALING_SPEED = 0.15;
// 最小缩放
export const MIN_SCALE_NUM = 0.1;
// 最大缩放
export const MAX_SCALE_NUM = 5;
// 原始尺寸的比例
export const FIT_TO_CONTAINER_BOUNDARY_PADDING = 72;
export const SCALE_OF_ORIGINAL_SIZE = 0.87;
export const VEC2_SCALE_OF_ORIGINAL_SIZE = 0.5;
// getNextScale
export const CURRENT_SCALE_OF_ORIGINAL_SIZE = 1;
export const DELTA_Y_SCALE = 0.01;
// 选择角度
export const ROTATION_RADIAN = -Math.PI / 2;
// 保留小数位数
export const TO_FIXED_NUMBER = 2;
// 小数放大
export const AMPLIFY_NUMBER = 100;
// 初始需求插件总数
export const PLUGIN_NUM = 4;
// 当插件个数为2
export const PLUGIN_NUM_2 = 2;
// 触摸板移动图片的延时
export const WHEEL_DELAY = 300;

export const TOOLBAR_DEFAULT = {
  pagination: true,
  scaling: true,
  rotation: true,
  download: true,
} satisfies ImageViewerToolbarProps;

export const TOOLBAR_FALSE = {
  pagination: false,
  scaling: false,
  rotation: false,
  download: false,
} satisfies ImageViewerToolbarProps;
