// @universe-design/react/src/image-viewer/ImageViewerImage.tsx
/* eslint-disable react/no-unknown-property */
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx, makeStylexStyles } from '@universe-design/stylex/react';
import * as React from 'react';
import { forwardRef, useEventCallback, useMergedRefs } from '@universe-design/react-utilities';
import { useImage_unstable } from '../shared/components/Image';
import { useImageViewerContext } from './ImageViewerContext';
import LoadfailFilled from '@universe-design/icons-react/LoadfailFilled';
import stylex from '@universe-design/stylex';
import { resolveSlot } from '../util/slot';
import { composeEventHandlers } from '../util/function';
import { Loading } from '../loading';
import { mat2d, vec2 } from 'gl-matrix';
import { usePointerHandler } from './usePointerHandler';
import { CURRENT_SCALE_OF_ORIGINAL_SIZE, DELTA_Y_SCALE } from './constants';
import { useImageViewerMobileScaling } from './useImageViewerMobileScaling';
import { ImageViewerCloseAction } from './useImageViewerVisible';

const useImageViewerImageStylexStyles = makeStylexStyles(({ prefix }: { prefix: string }) => ({
  image: stylex.create(
    {
      pointerEvents: 'auto',
      userSelect: 'none',
      touchAction: 'none',
      transformOrigin: 'left top',
      '-webkit-user-drag': 'none',
    },
    prefix,
  ),
  imageError: stylex.create({ display: 'none' }, `${prefix}-error`),
  loading: stylex.create(
    {
      position: 'absolute',
      top: '50%',
      left: '50%',
      transform: 'translate(-50%, -50%)',
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
    },
    `${prefix}__loading`,
  ),
  error: stylex.create(
    {
      position: 'absolute',
      top: '50%',
      left: '50%',
      transform: 'translate(-50%, -50%)',
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
      width: '120px',
      height: '120px',
      background: $token.N100,
      border: `1px solid ${$token['line-border-card']}`,
      borderRadius: $token['border-radius-xs'],
    },
    `${prefix}__error`,
  ),
  errorIcon: stylex.create(
    {
      color: $token.N400,
      fontSize: '48px',
    },
    `${prefix}__error__icon`,
  ),
  hideImage: stylex.create(
    {
      opacity: 0,
    },
    `${prefix}__hide`,
  ),
}));

const useImageViewerImage = () => {
  const { initTransform, setMatrixByPoints, scaleByOrigin, setDialogVisible } = useImageViewerContext();
  const [isImageVisible, setIsImageVisible] = React.useState(false);
  const [isPressing, setIsPressing] = React.useState(false);
  const [prevSrc, setPrevSrc] = React.useState('');

  /**
   * Translation
   *
   * - 按住拖拽（跨端同构，单点，onPointerDown + Move + Up）
   */
  const singleTouchTranslationHandler = usePointerHandler({
    down: () => {
      setIsPressing(true);
    },
    move: (ev, pointerIds) => {
      const lastEvent = pointerIds.get(ev.pointerId);
      if (pointerIds.size === 1 && lastEvent) {
        // 单点触控
        const { clientX, clientY } = lastEvent;

        setMatrixByPoints(point => {
          const translation = mat2d.fromTranslation(mat2d.create(), [ev.clientX - clientX, ev.clientY - clientY]);
          return vec2.transformMat2d(vec2.create(), point, translation);
        });
      }
    },
    up: () => {
      setIsPressing(false);
    },
    tap: event => {
      setDialogVisible(false, {
        visible: false,
        event,
        action: ImageViewerCloseAction.Image,
      });
    },
  });

  /**
   * Translation
   *
   * - 触控板滚动（onWheel）
   * - 鼠标滚轮（onWheel）
   */
  React.useEffect(() => {
    const handleWheel = (event: WheelEvent) => {
      const { metaKey, ctrlKey } = event;
      if (metaKey || ctrlKey) {
        return;
      }
      setMatrixByPoints(point => {
        const translation = mat2d.fromTranslation(mat2d.create(), [-event.deltaX, -event.deltaY]);
        return vec2.transformMat2d(vec2.create(), point, translation);
      });
    };
    document.addEventListener('wheel', handleWheel);
    return () => {
      document.removeEventListener('wheel', handleWheel);
    };
  });

  /**
   * Scaling
   * - 触控板 pinch（onWheel ctrl）
   * - 触控板 + meta（onWheel + meta）
   * - 鼠标滚轮 + ctrl（onWheel ctrl）
   */
  React.useEffect(() => {
    const handleWheel = (event: WheelEvent) => {
      event.preventDefault();
      const { deltaY, ctrlKey, metaKey } = event;
      if (metaKey || ctrlKey) {
        /**
         * 1. ctrl 或者 cmd 加 wheel 进行缩放
         * 2. 触控板 pinch 缩放，浏览器默认会加上 ctrlKey true
         */
        scaleByOrigin({
          origin: vec2.fromValues(event.clientX, event.clientY),
          getNextScale: currentScale => currentScale * (CURRENT_SCALE_OF_ORIGINAL_SIZE - deltaY * DELTA_Y_SCALE),
        });
        return;
      }
      return;
    };
    document.addEventListener('wheel', handleWheel, { passive: false });
    return () => {
      document.removeEventListener('wheel', handleWheel);
    };
  }, [scaleByOrigin]);

  const mobileScalingHandler = useImageViewerMobileScaling();

  const imageProps = {
    onLoad: () => {
      initTransform({ fitToContainer: true });
      setIsImageVisible(true);
    },
    onPointerDown: useEventCallback(composeEventHandlers(singleTouchTranslationHandler, mobileScalingHandler)),
  } satisfies React.ComponentProps<'img'>;

  return {
    slotsProps: {
      image: imageProps,
    },
    state: {
      isImageVisible,
      setIsImageVisible,
      prevSrc,
      setPrevSrc,
      isPressing,
    },
  };
};

export const ImageViewerImage = forwardRef<HTMLImageElement, ImageViewerImageExoticProps>((props, ref) => {
  const { errorFallback } = props;
  const { rootCssScope, imageRef, image, isImageContained } = useImageViewerContext();

  const {
    state: { status },
    slotsProps: useImageSlotsProps,
  } = useImage_unstable(image, useMergedRefs(ref, imageRef));
  const { slotsProps: useImageViewerImageSlotsProps, state: useImageViewerImageState } = useImageViewerImage();
  const { isImageVisible, setIsImageVisible, prevSrc, setPrevSrc, isPressing } = useImageViewerImageState;
  const styles = useImageViewerImageStylexStyles({
    prefix: rootCssScope('__image'),
  });

  if (image.src !== prevSrc) {
    setIsImageVisible(false);
    setPrevSrc(image.src);
  }

  // 获取光标形状
  const cursor = (() => {
    if (isImageContained) {
      return 'zoom-out';
    }
    if (isPressing) {
      return 'grabbing';
    }
    return 'grab';
  })();

  return (
    <React.Fragment>
      <img
        {...useImageSlotsProps.root}
        {...useImageViewerImageSlotsProps.image}
        sx={[
          useImageSlotsProps.root.sx,
          styles.image,
          status === 'error' && styles.imageError,
          !isImageVisible && styles.hideImage,
        ]}
        style={{ cursor, ...useImageSlotsProps.root.style }}
        alt={image?.alt ?? 'preview'}
        onLoad={composeEventHandlers(useImageSlotsProps.root.onLoad, useImageViewerImageSlotsProps.image.onLoad)}
      />
      {status === 'error' &&
        resolveSlot(errorFallback, {
          defaultNode: (
            <div sx={[styles.error]}>
              <LoadfailFilled className={stylex(styles.errorIcon)} />
            </div>
          ),
        })}
      {status === 'loading' && (
        <div sx={[styles.loading]}>
          <Loading size="large" />
        </div>
      )}
    </React.Fragment>
  );
});

export type ImageViewerImageProps = React.ComponentPropsWithoutRef<typeof ImageViewerImage>;

export interface ImageViewerImageExoticProps {
  /**
   * 如果加载失败显示的节点
   */
  errorFallback?: React.ReactNode;
}
