// @universe-design/react/src/image-viewer/ImageViewerToolbar.tsx
/* eslint-disable react/no-unknown-property */
/* eslint-disable react/jsx-no-leaked-render */
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx, makeStylexStyles } from '@universe-design/stylex/react';
import stylex from '@universe-design/stylex';
import { useImageViewerContext } from './ImageViewerContext';
import { forwardHTMLRef } from '../util/forwardRef';
import { renderSlotNode, resolveSlot } from '../shared/slot';
import { Space } from '../space';
import type { ImageViewerToolbarProps } from './useImageViewerToolbar';
import { Pagination } from './components/Pagination';
import { Scaling } from './components/Scaling';
import { Rotation } from './components/Rotation';
import { Download } from './components/Download';

import './style';

const useImageViewerToolbarStylexStyles = makeStylexStyles(({ prefix }: { prefix: string }) => ({
  toolbar: stylex.create(
    {
      userSelect: 'none',
      position: 'absolute',
      bottom: '16px',
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
      height: '48px',
      width: 'max-content',
      backgroundColor: $token['udtoken-imageview-toolbox-bg'],
      color: $token['static-white'],
      borderRadius: $token['border-radius-xs'],
      left: '50%',
      transform: 'translateX(-50%)',
      pointerEvents: 'auto',
      padding: '0 12px',
    },
    prefix,
  ),
  toolbarDivider: stylex.create(
    {
      width: '1px',
      height: '20px',
      background: $token['udtoken-dropdown-button-divider'],
    },
    `${prefix}__divider`,
  ),
}));
export const ImageViewerToolbar = forwardHTMLRef<ImageViewerToolbarProps, 'div'>((props, ref) => {
  const { pagination, scaling, rotation, download: downloadProp, children, ...htmlProps } = props;
  const { rootCssScope } = useImageViewerContext();
  const styles = useImageViewerToolbarStylexStyles({
    prefix: rootCssScope('__toolbar'),
  });

  const downloadNode = renderSlotNode(
    resolveSlot(downloadProp === true ? {} : downloadProp === false ? null : downloadProp, {
      elementType: Download,
      renderByDefault: true,
    }),
  );

  return (
    <div sx={[styles.toolbar]} {...htmlProps} ref={ref}>
      <Space size={24} split={<div className={stylex(styles.toolbarDivider)} />}>
        {pagination && <Pagination />}
        {scaling && <Scaling />}
        {(rotation || downloadNode) && (
          <Space size={12}>
            {rotation && <Rotation />}
            {downloadNode}
          </Space>
        )}
        {children}
      </Space>
    </div>
  );
});
