// @universe-design/react/src/image-viewer/ImageViewerDialog.tsx
/* eslint-disable react/no-unknown-property */
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx } from '@universe-design/stylex/react';
import * as React from 'react';
import stylex from '@universe-design/stylex';
import type { DialogProps } from '../dialog/dialog';
import { Dialog, DialogCancelType, DialogLayoutMode } from '../dialog/dialog';
import type { UseImageViewerVisibleProps } from './useImageViewerVisible';
import { ImageViewerCloseAction } from './useImageViewerVisible';
import { IconButton } from '../button';
import CloseOutlined from '@universe-design/icons-react/CloseOutlined';
import { useImageViewerContext } from './ImageViewerContext';

import './style';

export const ImageViewerDialog = React.forwardRef<
  HTMLDivElement,
  {
    /**
     * 是否显示右上角关闭按钮
     * @defaultValue true
     */
    hasCloseButton?: boolean;
    /**
     * 点击遮罩是否允许关闭
     * @defaultValue true
     */
    maskClosable?: boolean;
  } & Pick<DialogProps, 'zIndex' | 'keyboardClosable'> &
    UseImageViewerVisibleProps
>((props, ref) => {
  // TODO 原则上来说这块也可以有一个 useImageViewerDialogProps 的 hooks 专门预处理 props，我们的 render 函数体里要尽可能干净
  const { zIndex, maskClosable = true, keyboardClosable, hasCloseButton = true, children } = props;

  const { rootCssScope, containerRef, dialogVisible, setDialogVisible } = useImageViewerContext();

  return (
    <DialogLayoutMode mode="2.18.0">
      <Dialog
        ref={ref}
        zIndex={zIndex}
        safeMargin={{ hmax: 0, hmin: 0, top: 0, bottom: 0 }}
        width="100%"
        wrapperClassName={rootCssScope('__dialog__wrap')}
        destroyOnClose
        keyboardClosable={keyboardClosable}
        overlayClosable={maskClosable}
        overlayClassName={rootCssScope('__dialog__mask')}
        visible={dialogVisible}
        // dialog mask 和 esc 关闭
        onClose={(e, { cancelType }) => {
          if (cancelType === DialogCancelType.Esc) {
            setDialogVisible(false, {
              visible: false,
              event: e as React.KeyboardEvent,
              action: ImageViewerCloseAction.Esc,
            });
            return;
          }
          if (cancelType === DialogCancelType.Mask) {
            setDialogVisible(false, {
              visible: false,
              event: e as React.MouseEvent,
              action: ImageViewerCloseAction.Mask,
            });
            return;
          }
        }}
      >
        <div
          ref={containerRef}
          sx={[
            stylex.create(
              {
                width: '100%',
                height: '100%',
                position: 'absolute',
              },
              rootCssScope(),
            ),
          ]}
        >
          {children}
          {hasCloseButton && (
            <IconButton
              size={20}
              icon={<CloseOutlined />}
              className={rootCssScope('__close-button')}
              onClick={ev => {
                setDialogVisible(false, {
                  visible: false,
                  event: ev,
                  action: ImageViewerCloseAction.CloseButton,
                });
              }}
            />
          )}
        </div>
      </Dialog>
    </DialogLayoutMode>
  );
});

export type ImageViewerDialogProps = React.ComponentPropsWithoutRef<typeof ImageViewerDialog>;
