// @universe-design/react/src/image-viewer/useImageViewerTransform.ts
/**
 * ================================
 *
 * useImageViewerTransform
 *
 * ================================
 */
import * as React from 'react';
import { mat2d, vec2 } from 'gl-matrix';
import { useEventCallback, useResizeObserver, useLatestState, getSize } from '@universe-design/react-utilities';
import type { Points } from './transform';
import {
  clampPoints,
  getMatRotationRadians,
  getMatScale,
  getTransformBetweenTwoRects,
  isPointsInRectangleRegion,
  rectToPoints,
} from './transform';
import { FIT_TO_CONTAINER_BOUNDARY_PADDING, MAX_SCALE_NUM, MIN_SCALE_NUM } from './constants';
import { isFunction } from '../util/assert';

/**
 * 或者图片的原始大小
 */
const getImageOriginSize = (el: HTMLImageElement | null) => ({
  width: el?.naturalWidth ?? 0,
  height: el?.naturalHeight ?? 0,
});

/**
 * 与图片变换相关的逻辑
 */
export const useImageViewerTransform = (props: {
  /**
   * 最大缩放比例
   * @defaultValue 5
   */
  maxScale?: number;
  /**
   * 最小缩放比例
   * @defaultValue 0.1
   */
  minScale?: number;
}) => {
  const { minScale = MIN_SCALE_NUM, maxScale = MAX_SCALE_NUM } = props;
  const imageRef = React.useRef<HTMLImageElement>(null);
  const [matrix, _setMatrix, matrixRef] = useLatestState(mat2d.create());
  const [containerSize, setContainerSize] = React.useState({
    width: 0,
    height: 0,
  });
  const containerRef = useResizeObserver<HTMLDivElement>(entry => {
    const imageOriginalRect = getImageOriginalRect();
    const imageOriginalPoints = rectToPoints(imageOriginalRect);
    const nextPoints = imageOriginalPoints.map(p => vec2.transformMat2d(vec2.create(), p, matrixRef.current)) as Points;
    const containerRect = getContainerRect(entry);
    const clampedMat = getClampedMatrix({
      currentPoints: imageOriginalPoints,
      nextPoints,
      containerRect,
    });
    clampedMat && setMatrix(clampedMat);
    setContainerSize(getSize(entry.target, entry));
  });

  /**
   * 获取夹过的变换矩阵
   */
  const getClampedMatrix = useEventCallback(
    ({
      currentPoints,
      nextPoints,
      containerRect,
    }: {
      currentPoints: Points;
      nextPoints: Points;
      containerRect: DOMRect;
    }) => {
      const clampedPoints = clampPoints(nextPoints, {
        containerRect,
      });
      return getTransformBetweenTwoRects(currentPoints, clampedPoints);
    },
  );

  /**
   * Matrix
   */
  const getMatrix = useEventCallback(() => matrixRef.current);

  const setMatrix = useEventCallback<React.Dispatch<mat2d>>(mat => {
    const image = imageRef.current;
    if (image) {
      // 我们使用原生 dom 操作跳过 react 直接调整图片变换
      image.style.transform = `matrix(${mat.join(', ')})`;
    }
    _setMatrix(mat);
  });

  /**
   * Image
   */
  const [imageOriginalSize, setImageOriginalSize] = React.useState({
    width: 0,
    height: 0,
  });

  const getImageOriginalRect = useEventCallback((): DOMRect => {
    const { width, height } = getImageOriginSize(imageRef.current);
    return new DOMRect(0, 0, width, height);
  });

  const setMatrixByPoints = useEventCallback((pointsOrGetNextPoint: Points | ((point: vec2) => vec2)) => {
    const originalPoints = rectToPoints(getImageOriginalRect());
    const currentMat = getMatrix();
    const nextPoints = isFunction(pointsOrGetNextPoint)
      ? (originalPoints.map(point =>
          pointsOrGetNextPoint(vec2.transformMat2d(vec2.create(), point, currentMat)),
        ) as Points)
      : pointsOrGetNextPoint;
    const containerRect = getContainerRect();
    const clampedMat = getClampedMatrix({
      currentPoints: originalPoints,
      nextPoints,
      containerRect,
    });
    setMatrix(clampedMat);
  });

  /**
   * Container
   */
  const getContainerRect = useEventCallback((entry?: ResizeObserverEntry) => {
    const container = containerRef.current;
    if (!container) {
      return new DOMRect();
    }
    const containerSize = getSize(container, entry);
    const containerRect = new DOMRect(0, 0, containerSize.width, containerSize.height);
    return containerRect;
  });

  const initTransform = useEventCallback((options: { fitToContainer?: boolean } = {}) => {
    const { fitToContainer = true } = options;
    const imageOriginalRect = getImageOriginalRect();
    const container = containerRef.current;
    if (!imageOriginalRect || !container) {
      return;
    }
    // 用 layout size 而不是 bounding rect 可以消除动画带来的影响
    const { clientWidth: containerWidth, clientHeight: containerHeight } = container;

    const nextImageRect = (() => {
      let width = imageOriginalRect.width;
      let height = imageOriginalRect.height;
      if (fitToContainer) {
        // 算出 x y 完全卡到 boundary 时的缩放比，取小的那个可以保证是 contain 的效果
        const scale = Math.min(
          Math.max(
            Math.min(
              (containerWidth - FIT_TO_CONTAINER_BOUNDARY_PADDING * 2) / imageOriginalRect.width,
              (containerHeight - FIT_TO_CONTAINER_BOUNDARY_PADDING * 2) / imageOriginalRect.height,
              1,
            ),
            minScale,
          ),
          maxScale,
        );
        width = imageOriginalRect.width * scale;
        height = imageOriginalRect.height * scale;
      }
      // TODO 这里其实有问题，如果 container x 和 y 不在 0 就不对了
      return new DOMRect((containerWidth - width) / 2, (containerHeight - height) / 2, width, height);
    })();
    setImageOriginalSize(imageOriginalRect);
    setMatrix(getTransformBetweenTwoRects(rectToPoints(imageOriginalRect), rectToPoints(nextImageRect)));
  });

  const getScale = useEventCallback(() => getMatScale(matrix));

  /**
   * 按中心缩放
   */
  const scaleByOrigin = useEventCallback(
    ({
      origin: originParam,
      getNextScale,
    }: {
      /**
       * 缩放原点
       */
      origin?: vec2;
      /**
       * 缩放尺度
       */
      getNextScale: (currentScale: number) => number;
    }) => {
      const origin = (() => {
        if (originParam) {
          return originParam;
        }
        const containerRect = getContainerRect();
        return vec2.fromValues(
          (containerRect.x + containerRect.width) / 2,
          (containerRect.y + containerRect.height) / 2,
        );
      })();
      const currentScale = getScale();
      const nextScale = Math.max(minScale, Math.min(getNextScale(currentScale), maxScale)); // Limit the scale

      if (nextScale === currentScale) {
        return;
      }

      const scaleMat = mat2d.fromScaling(mat2d.create(), [nextScale / currentScale, nextScale / currentScale]);

      setMatrixByPoints(absolutePoint => {
        const point = vec2.subtract(vec2.create(), absolutePoint, origin); // 基于 origin 的坐标
        let nextPoint = vec2.create();
        nextPoint = vec2.transformMat2d(vec2.create(), point, scaleMat);
        nextPoint = vec2.add(vec2.create(), nextPoint, origin);
        return nextPoint;
      });
    },
  );

  /**
   * 这里用了一些场外信息，我们的 image 只会做 scaleX 等于 scaleY 的等比缩放
   */
  const scale = getMatScale(matrix);

  const rotationRadians = getMatRotationRadians(matrix);

  /**
   * 判断图片尺寸是否大于容器尺寸
   *  */
  const isImageContained = React.useMemo(() => {
    if (typeof DOMRect === 'undefined') {
      return false;
    }
    const imageOriginalRect = new DOMRect(0, 0, imageOriginalSize.width, imageOriginalSize.height);
    const originalPoints = [
      ...rectToPoints(imageOriginalRect),
      vec2.fromValues(imageOriginalRect.x + imageOriginalRect.width, imageOriginalRect.y + imageOriginalRect.height),
    ];
    const transformedPoints: Points = [vec2.create(), vec2.create(), vec2.create()];
    for (const [i, transformedPoint] of transformedPoints.entries()) {
      vec2.transformMat2d(transformedPoint, originalPoints[i], matrix);
    }
    return isPointsInRectangleRegion(new DOMRect(0, 0, containerSize.width, containerSize.height), transformedPoints);
  }, [containerSize.height, containerSize.width, imageOriginalSize.height, imageOriginalSize.width, matrix]);

  return {
    state: {
      getMatrix,
      setMatrix,
      setMatrixByPoints,
      imageRef,
      containerRef,
      getContainerRect,
      initTransform,
      getScale,
      scale,
      rotationRadians,
      scaleByOrigin,
      minScale,
      maxScale,
      /**
       * 判断图片区域是否超出容器区域
       */
      isImageContained,
    },
  };
};

export type UseImageViewerTransformProps = Parameters<typeof useImageViewerTransform>[0];
