// @universe-design/react/src/image-viewer/useImageViewerVisible.ts
import { useControlledState } from '@universe-design/react-utilities';
import { isFunction } from '../util';
import { useEventCallback } from '../hooks';

/**
 * ================================
 *
 * useImageViewerVisible
 *
 * ================================
 */

/**
 * 与图片查看器弹层控制相关的数据逻辑
 */
export const useImageViewerVisible = (props: {
  /**
   * 图片查看器可见性相关属性
   */
  visible?: ImageViewerVisibleProps;
}) => {
  const { visible } = props;
  const [dialogVisible, setDialogVisible] = useControlledState(visible?.defaultValue ?? false, visible?.value);
  const setDialogVisibleWithMeta = useEventCallback<NonNullable<ImageViewerVisibleProps['onChange']>>((value, meta) => {
    setDialogVisible(value);
    if (!visible) {
      return;
    }
    const { onChange } = visible;
    isFunction(onChange) && onChange(value, meta);
  });

  return {
    state: {
      dialogVisible,
      setDialogVisible: setDialogVisibleWithMeta,
    },
  };
};

/**
 * 图片查看器可见性相关属性
 */
export interface ImageViewerVisibleProps {
  /**
   * 图片查看器是否可见的受控值
   */
  value?: boolean;
  /**
   * 图片查看器是否可见的默认值
   * @defaultValue false
   */
  defaultValue?: boolean;
  /**
   * 图片查看器是否可见的回调函数
   */
  onChange?: (value: false, meta: ImageViewerVisibleOnChangeMeta) => void;
}

/** 关闭图片查看器的参数 */
export type ImageViewerVisibleOnChangeMeta =
  | {
      visible: false;
      event: React.MouseEvent;
      action: ImageViewerCloseAction.Mask;
    }
  | {
      visible: false;
      event: React.MouseEvent;
      action: ImageViewerCloseAction.CloseButton;
    }
  | {
      visible: false;
      event: React.KeyboardEvent;
      action: ImageViewerCloseAction.Esc;
    }
  | {
      visible: false;
      event: PointerEvent;
      action: ImageViewerCloseAction.Image;
    };

/**
 * 图片查看器关闭动作
 */
export enum ImageViewerCloseAction {
  /** 点击 mask 关闭 */
  Mask = 'mask',
  /** 是否显示右上角关闭按钮 */
  CloseButton = 'closeButton',
  /** 按 Esc 键关闭预览 */
  Esc = 'esc',
  /** 点击 image  关闭预览 */
  Image = 'image',
}

export type UseImageViewerVisibleProps = Parameters<typeof useImageViewerVisible>[0];
