// @universe-design/react/src/image-viewer/useImageViewerImages.ts
import type * as React from 'react';
import { useControlledState } from '@universe-design/react-utilities';

/**
 * ================================
 *
 * useImageViewerImages
 *
 * ================================
 */

/**
 * 与图片查看器图片控制相关的数据逻辑
 */
export const useImageViewerImages = (props: UseImageViewerImagesProps) => {
  const { images, index } = props;

  const [currentIndex, setCurrentIndex] = useControlledState(index?.defaultValue ?? 0, index?.value, {
    onChange: index?.onChange,
  });

  const image = images[currentIndex];

  return {
    state: { currentIndex, setCurrentIndex, image },
  };
};

export interface UseImageViewerImagesProps {
  /**
   * 图片集合
   */
  images: ImageProps[];
  /**
   * 图片下标
   * @defaultValue `{ defaultValue: 0 }`
   */
  index?: ImageViewerIndexProps;
}

/**
 * 图片查看器中的图片
 */
export interface ImageProps extends React.ComponentPropsWithoutRef<'img'> {
  /**
   * 图片的src
   */
  src: string;
  /**
   * 图片的alt，下载图片的名称
   * @defaultValue `${new Date().getTime()}.jpg`
   */
  alt?: string;
}

/**
 * 图片查看器的图片下标
 */
export interface ImageViewerIndexProps {
  /**
   * 当前展示的图片下标，开发时注意不能超过数组大小
   */
  value?: number;
  /**
   * 默认展示的图片下标
   * @defaultValue 0
   */
  defaultValue?: number;
  /**
   * 更改图片下标的调用
   */
  onChange?: (value: number) => void;
}
