// @universe-design/react/src/image-viewer/usePointerHandler.tsx
import { useEventCallback, useConstant } from '@universe-design/react-utilities';

type PointerHandlerFn = (ev: PointerEvent, prevPointerIds: Map<number, PointerEvent>) => any;

const POINTER_TAP_THRESHOLD = 300;
const POINTER_TAP_MOVE_THRESHOLD = 5;

/**
 * 处理多点触控场景的
 *
 * @privateRemarks
 *
 * 返回的处理函数需要传递给 PointerDown React Event Handler
 */
export const usePointerHandler = (pointerHandler: {
  down?: PointerHandlerFn;
  move?: PointerHandlerFn;
  up?: PointerHandlerFn;
  tap?: PointerHandlerFn;
}) => {
  const pointerIds = useConstant(() => new Map<number, PointerEvent>());
  const startEvents = useConstant(
    () =>
      new Map<
        number,
        {
          time: number;
          x: number;
          y: number;
          moved: boolean;
        }
      >(),
  );
  const { down, move, up, tap } = pointerHandler;
  const hasHandler = down || move || tap;

  const handler = useEventCallback(_event => {
    const { nativeEvent } = _event;

    pointerIds.set(_event.pointerId, nativeEvent);

    // 记录按下事件的初始状态
    startEvents.set(_event.pointerId, {
      time: Date.now(),
      x: nativeEvent.clientX,
      y: nativeEvent.clientY,
      moved: false,
    });

    down?.(_event, new Map(pointerIds));

    if (pointerIds.size > 1) {
      return;
    }

    const handelPointerUpOrCancel = (ev: PointerEvent) => {
      pointerIds.delete(ev.pointerId);
      const prevPointerIds = new Map(pointerIds);
      up?.(ev, prevPointerIds);

      // 处理 tap 事件
      if (tap) {
        const startEvent = startEvents.get(ev.pointerId);
        if (startEvent) {
          const currentTime = Date.now();
          const timeDiff = currentTime - startEvent.time;
          if (timeDiff <= POINTER_TAP_THRESHOLD && !startEvent.moved) {
            tap(ev, prevPointerIds);
          }
        }
      }

      // 清理状态
      startEvents.delete(ev.pointerId);

      if (pointerIds.size === 0) {
        document.removeEventListener('pointerup', handelPointerUpOrCancel);
        document.removeEventListener('pointercancel', handelPointerUpOrCancel);
        document.removeEventListener('pointermove', handelPointerMove);
      }
    };

    const handelPointerMove = (ev: PointerEvent) => {
      // 检测移动阈值（5px）
      const startEvent = startEvents.get(ev.pointerId);
      if (startEvent) {
        const deltaX = Math.abs(ev.clientX - startEvent.x);
        const deltaY = Math.abs(ev.clientY - startEvent.y);
        if (deltaX > POINTER_TAP_MOVE_THRESHOLD || deltaY > POINTER_TAP_MOVE_THRESHOLD) {
          startEvent.moved = true;
        }
      }

      // clone 一份，以免用户把 internal 数据改坏掉
      const prevPointerIds = new Map(pointerIds);
      pointerIds.set(ev.pointerId, ev);
      move?.(ev, prevPointerIds);
    };

    document.addEventListener('pointerup', handelPointerUpOrCancel);
    document.addEventListener('pointercancel', handelPointerUpOrCancel);
    document.addEventListener('pointermove', handelPointerMove);
  });

  return hasHandler ? handler : undefined;
};
