// @universe-design/react-utilities/lib-esm/descendant/use-descendant.js
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/naming-convention */
/* eslint-disable @typescript-eslint/no-magic-numbers */
import React, { useRef, useState } from 'react';
import { useIsomorphicLayoutEffect as useLayoutEffect } from '../hooks/useIsomorphicLayoutEffect';
import { useForceUpdate } from '../hooks/useForceUpdate';
import { composeRefs } from '../hooks/composeRefs';
import { createContextHook } from '@universe-design/react-provider';
import { DescendantsManager } from './descendant';
import { useConstant } from '../hooks/useConstant';

/**
 * @internal
 * React hook that initializes the DescendantsManager
 */
function createUseDescendants() {
  return function () {
    const descendants = useConstant(() => new DescendantsManager());
    useLayoutEffect(() => () => descendants.destroy(), [descendants]);
    return descendants;
  };
}
function createUseDescendant(useDescendantsContext) {
  return function (options) {
    var _descendants$enabledI;
    const descendants = useDescendantsContext();
    const forceUpdate = useForceUpdate();
    const [index, setIndex] = useState(-1);
    const ref = useRef(null);
    // 只在 unmount 时执行一次
    useLayoutEffect(() => () => {
      if (!ref.current || !descendants) {
        return;
      }
      descendants.unregister(ref.current);
    }, [descendants]);
    // eslint-disable-next-line react-hooks/exhaustive-deps
    useLayoutEffect(() => {
      if (!ref.current) {
        return;
      }
      const dataIndex = Number(ref.current.dataset.index);
      if (index !== dataIndex && !Number.isNaN(dataIndex)) {
        setIndex(dataIndex);
      }
    });
    const register = React.useMemo(() => composeRefs(descendants === null || descendants === void 0 ? void 0 : descendants.register({
      forceUpdate
    }, options), ref), [descendants, forceUpdate, options]);
    return {
      descendants,
      index,
      enabledIndex: (_descendants$enabledI = descendants === null || descendants === void 0 ? void 0 : descendants.enabledIndexOf(ref.current)) !== null && _descendants$enabledI !== void 0 ? _descendants$enabledI : -1,
      register
    };
  };
}
/**
 * Create a generic typed descendant context
 *
 * @public
 */
export function createDescendantContext() {
  const Context = createContextHook('DescendantsContext', {
    strict: true
  });
  return [
  // context provider
  Context.Provider,
  // call this when you need to read from context
  Context.useContext,
  // descendants state information, to be called and passed to `ContextProvider`
  createUseDescendants(),
  // descendant index information
  createUseDescendant(Context.useContext)];
}
/**
 * Create a generic typed descendant context
 *
 * @public
 */
export function createDescendantContextOptional() {
  const Context = createContextHook('DescendantContext(Optional)', {
    strict: false,
    defaultValue: null
  });
  return [
  // context provider
  Context.Provider,
  // call this when you need to read from context
  Context.useContext,
  // descendants state information, to be called and passed to `ContextProvider`
  createUseDescendants(),
  // descendant index information
  createUseDescendant(Context.useContext)];
}
