// @universe-design/react/src/shared/components/popper/test-demo/test-placement.tsx
import * as React from 'react';
import { Space } from '@universe-design/react';
import { Popover, buttonWidth } from '../test-utils/components';

const spaceSize = 0;

export default () => (
  <div
    id="container"
    style={{
      padding: '0 0 0 300px',
    }}
  >
    <div style={{ height: 300 }} />
    <Space
      size={spaceSize}
      style={{
        display: 'flex',
        marginLeft: buttonWidth + spaceSize,
      }}
    >
      <Popover placement="top-start" />
      <Popover placement="top" />
      <Popover placement="top-end" />
    </Space>
    <Space
      size={spaceSize}
      direction="vertical"
      style={{
        display: 'flex',
        width: buttonWidth,
        marginTop: spaceSize,
        float: 'left',
      }}
    >
      <Popover placement="left-start" />
      <Popover placement="left" />
      <Popover placement="left-end" />
    </Space>
    <Space
      size={spaceSize}
      direction="vertical"
      style={{
        display: 'flex',
        width: buttonWidth,
        marginTop: spaceSize,
        marginLeft: buttonWidth * 4 + spaceSize * 4,
      }}
    >
      <Popover placement="right-start" />
      <Popover placement="right" />
      <Popover placement="right-end" />
    </Space>
    <Space
      size={spaceSize}
      style={{
        display: 'flex',
        marginLeft: buttonWidth + spaceSize,
        marginTop: spaceSize,
        clear: 'both',
      }}
    >
      <Popover placement="bottom-start" />
      <Popover placement="bottom" />
      <Popover placement="bottom-end" />
    </Space>
    <div style={{ height: 1200 }} />
  </div>
);

export const screenshot = {
  skip: true,
};
