// @universe-design/react/src/menu/internal-menu/SubMenu/PopupTrigger.tsx
import * as React from 'react';
import Trigger from '@universe-design/rc-trigger';
import clsx from 'clsx';
import raf from '@universe-design/rc-util/lib/raf';
import type { CSSMotionProps } from '@universe-design/rc-motion';
import { RcMenuContext } from '../context/MenuContext';
import { rcPlacements, rcPlacementsRtl } from '../placements';
import type { RcMenuMode } from '../interface';
import { getMotion } from '../utils/motionUtil';

const popupPlacementMap = {
  horizontal: 'bottomLeft',
  vertical: 'rightTop',
  'vertical-left': 'rightTop',
  'vertical-right': 'leftTop',
} as const;

export interface PopupTriggerProps {
  prefixCls: string;
  mode: RcMenuMode;
  visible: boolean;
  children: React.ReactElement;
  popup: React.ReactNode;
  popupClassName?: string;
  popupOffset?: number[];
  disabled: boolean;
  onVisibleChange: (visible: boolean) => void;
}

export default function PopupTrigger({
  prefixCls,
  visible,
  children,
  popup,
  popupClassName,
  popupOffset,
  disabled,
  mode,
  onVisibleChange,
}: PopupTriggerProps) {
  const {
    getPopupContainer,
    rtl,
    subMenuOpenDelay,
    subMenuCloseDelay,
    builtinPlacements,
    triggerSubMenuAction,
    forceSubMenuRender,

    // Motion
    motion,
    defaultMotions,
  } = React.useContext(RcMenuContext)!;

  const [innerVisible, setInnerVisible] = React.useState(false);

  const placement = rtl ? { ...rcPlacementsRtl, ...builtinPlacements } : { ...rcPlacements, ...builtinPlacements };

  const popupPlacement = popupPlacementMap[mode as keyof typeof popupPlacementMap];

  const targetMotion = getMotion(mode, motion, defaultMotions);

  const mergedMotion: CSSMotionProps = {
    ...targetMotion,
    leavedClassName: `${prefixCls}-hidden`,
    removeOnLeave: false,
    motionAppear: true,
  };

  // Delay to change visible
  const visibleRef = React.useRef<number>();

  React.useEffect(() => {
    visibleRef.current = raf(() => {
      setInnerVisible(visible);
    });

    return () => {
      raf.cancel(visibleRef.current!);
    };
  }, [visible]);

  return (
    <Trigger
      prefixCls={prefixCls}
      popupClassName={clsx(
        `${prefixCls}-popup`,
        {
          [`${prefixCls}-rtl`]: rtl,
        },
        popupClassName,
      )}
      stretch={mode === 'horizontal' ? 'minWidth' : undefined}
      getPopupContainer={getPopupContainer}
      builtinPlacements={placement}
      popupPlacement={popupPlacement}
      popupVisible={innerVisible}
      popup={popup}
      popupAlign={popupOffset && { offset: popupOffset }}
      action={disabled ? [] : triggerSubMenuAction && [triggerSubMenuAction]}
      mouseEnterDelay={subMenuOpenDelay}
      mouseLeaveDelay={subMenuCloseDelay}
      onPopupVisibleChange={onVisibleChange}
      forceRender={forceSubMenuRender}
      popupMotion={mergedMotion}
    >
      {children}
    </Trigger>
  );
}
