// @universe-design/react/es/pagination/Pagination.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.u1k6fduh{cursor:pointer;}
import u12o6wmt from "./../.stylex/u12o6wmt";
//.u1fssmi4{color:@icon-n2;}
import u1rkqst1 from "./../.stylex/u1rkqst1";
import React from 'react';
import RcPagination from './internal';
import clsx from 'clsx';
import LeftBoldOutlined from '@universe-design/icons-react/LeftBoldOutlined';
import RightBoldOutlined from '@universe-design/icons-react/RightBoldOutlined';
import UpTopOutlined from '@universe-design/icons-react/UpTopOutlined';
import MoreOutlined from '@universe-design/icons-react/MoreOutlined';
import stylex from '@universe-design/stylex';
import { PaginationSelect } from './PaginationSelect';
import { useBreakpoint } from '../grid/hooks/useBreakpoint';
import { useCssScope, useDirection, useLocaleReceiver } from '../config-provider/hooks';
import { warning, compatibleDevWarning, deprecatedWarning } from '../util';
import { usePaginationContext } from './PaginationContext';
import './style';

/**
 * @docgen
 */

const usePaginationCssScope = () => useCssScope('pagination');
const sxEllipsisIcon = {
  "props": {
    "sj55zd": u1rkqst1
  }
};
const sxEllipsis = {
  "props": {
    "Bceei9c": u12o6wmt
  }
};
const Ellipsis = _ref => {
  let {
    position
  } = _ref;
  const prefixCls = usePaginationCssScope()('');
  const {
    ellipsisReadOnly
  } = usePaginationContext();
  if (ellipsisReadOnly) {
    return /*#__PURE__*/React.createElement(MoreOutlined, {
      className: stylex(sxEllipsisIcon)
    });
  }
  return /*#__PURE__*/React.createElement("a", {
    className: clsx(`${prefixCls}-item-link`, stylex(sxEllipsis))
  }, /*#__PURE__*/React.createElement("div", {
    className: `${prefixCls}-item-container`
  }, /*#__PURE__*/React.createElement(UpTopOutlined
  // FIXME 修改为正常的 icon
  , {
    style: {
      transform: position === 'start' ? 'rotate(-90deg)' : 'rotate(90deg)'
    },
    className: `${prefixCls}-item-link-icon`
  }), /*#__PURE__*/React.createElement("span", {
    className: `${prefixCls}-item-ellipsis`
  }, /*#__PURE__*/React.createElement(MoreOutlined, null))));
};

/**
 * @public
 */
export const Pagination = _ref2 => {
  let {
    className,
    size,
    itemRender,
    simple,
    showQuickJumper,
    showSizeChanger,
    showLessItems,
    ...restProps
  } = _ref2;
  process.env.NODE_ENV !== "production" ? warning(typeof simple !== 'boolean', compatibleDevWarning('Pagination', ['simple'])) : void 0;
  process.env.NODE_ENV !== "production" ? warning(typeof showLessItems === 'undefined', deprecatedWarning('Pagination', 'showLessItems', ['siblingCount', 'boundaryCount'])) : void 0;
  const {
    xs
  } = useBreakpoint();
  const direction = useDirection();
  const prefixCls = useCssScope('pagination')('');
  const selectPrefixCls = useCssScope('select')('');
  const locale = useLocaleReceiver('pagination');
  const isSmall = size === 'small' || Boolean(xs && !size && restProps.responsive);
  const extendedClassName = clsx({
    mini: isSmall,
    [`${prefixCls}-rtl`]: direction === 'rtl'
  }, className);
  const mergedShowQuickJumper = showQuickJumper !== null && showQuickJumper !== void 0 ? showQuickJumper : false;
  const mergedShowSizeChanger = // eslint-disable-next-line @typescript-eslint/no-magic-numbers
  showSizeChanger !== null && showSizeChanger !== void 0 ? showSizeChanger : Boolean(!isSmall && typeof restProps.total === 'number' && restProps.total > 50);
  return /*#__PURE__*/React.createElement(RcPagination, _extends({}, restProps, {
    prefixCls: prefixCls,
    selectPrefixCls: selectPrefixCls
  }, (() => {
    let prevIcon = /*#__PURE__*/React.createElement("button", {
      className: `${prefixCls}-item-link`,
      type: "button",
      tabIndex: -1
    }, /*#__PURE__*/React.createElement(LeftBoldOutlined, null));
    let nextIcon = /*#__PURE__*/React.createElement("button", {
      className: `${prefixCls}-item-link`,
      type: "button",
      tabIndex: -1
    }, /*#__PURE__*/React.createElement(RightBoldOutlined, null));
    let jumpPrevIcon = /*#__PURE__*/React.createElement(Ellipsis, {
      position: "start"
    });
    let jumpNextIcon = /*#__PURE__*/React.createElement(Ellipsis, {
      position: "end"
    });
    // change arrows direction in right-to-left direction
    if (direction === 'rtl') {
      [prevIcon, nextIcon] = [nextIcon, prevIcon];
      [jumpPrevIcon, jumpNextIcon] = [jumpNextIcon, jumpPrevIcon];
    }
    return {
      prevIcon,
      nextIcon,
      jumpPrevIcon,
      jumpNextIcon
    };
  })(), {
    className: extendedClassName,
    selectComponentClass: PaginationSelect,
    locale: locale,
    itemRender: itemRender,
    showQuickJumper: mergedShowQuickJumper,
    showSizeChanger: mergedShowSizeChanger
  }));
};
if (process.env.NODE_ENV !== "production") {
  Pagination.displayName = 'Pagination';
}
