// @universe-design/react/es/pagination/usePagination.js
import * as React from 'react';
import clamp from 'lodash/clamp';
import maxBy from 'lodash/maxBy';
import { useControlledState } from '@universe-design/react-utilities';
import { isFunction, isNonNullable } from '../util';
/**
 * utils
 */
const getPageCount = _ref => {
  let {
    total,
    pageSize
  } = _ref;
  return Math.ceil(total / pageSize);
};
const clampPageWithRange = _ref2 => {
  let {
    page,
    minPage,
    maxPage
  } = _ref2;
  return clamp(page, minPage, maxPage);
};

/**
 * usePagination
 *
 * 抽象的分页数据处理逻辑
 */
export const usePagination = props => {
  const {
    total: totalProp = 0,
    current: currentProp,
    defaultCurrent = 1,
    pageSize: pageSizeProp,
    defaultPageSize = 10,
    siblingCount: siblingCountProp,
    boundaryCount: boundaryCountProp,
    showEllipses: showEllipsesProp = true,
    onPageChange,
    onPageSizeChange,
    disabled = false,
    enableSiblingWindowBoundary = true
  } = props;

  /**
   * ================================================
   *
   * Total Prop
   *
   * ================================================
   */
  const total = clamp(totalProp, 0, Infinity);

  /**
   * ================================================
   *
   * Sibling Count Prop
   *
   * ================================================
   */
  const siblingCount = React.useMemo(() => {
    if (!isNonNullable(siblingCountProp)) {
      return {
        start: 1,
        end: 1
      };
    }
    if (typeof siblingCountProp === 'number') {
      return {
        start: siblingCountProp,
        end: siblingCountProp
      };
    }
    return siblingCountProp;
  }, [siblingCountProp]);

  /**
   * ================================================
   *
   * Boundary Count Prop
   *
   * ================================================
   */
  const boundaryCount = React.useMemo(() => {
    if (!isNonNullable(boundaryCountProp)) {
      return {
        start: 1,
        end: 1
      };
    }
    if (typeof boundaryCountProp === 'number') {
      return {
        start: boundaryCountProp,
        end: boundaryCountProp
      };
    }
    return boundaryCountProp;
  }, [boundaryCountProp]);

  /**
   * ================================================
   *
   * Start End Ellipsis Prop
   *
   * ================================================
   */
  const {
    startEllipsisCount,
    endEllipsisCount,
    ellipsisThreshold,
    ellipsisReadOnly
  } = React.useMemo(() => {
    const defaultEllipsisThreshold = 2;
    const defaultEllipsisReadOnly = false;
    if (!showEllipsesProp) {
      return {
        startEllipsisCount: 0,
        endEllipsisCount: 0,
        ellipsisThreshold: defaultEllipsisThreshold,
        ellipsisReadOnly: defaultEllipsisReadOnly
      };
    }
    if (showEllipsesProp === true) {
      return {
        startEllipsisCount: 1,
        endEllipsisCount: 1,
        ellipsisThreshold: defaultEllipsisThreshold,
        ellipsisReadOnly: defaultEllipsisReadOnly
      };
    }
    const {
      threshold = defaultEllipsisThreshold,
      readOnly = defaultEllipsisReadOnly
    } = showEllipsesProp;
    return {
      startEllipsisCount: 1,
      endEllipsisCount: 1,
      ellipsisThreshold: threshold,
      ellipsisReadOnly: readOnly
    };
  }, [showEllipsesProp]);

  /**
   * ================================================
   *
   * PageSize PageCount
   *
   * ================================================
   */
  const clampPageSize = pageSize => clamp(pageSize, 1, Infinity);
  const [pageSize, setPageSize] = useControlledState(clampPageSize(defaultPageSize), isNonNullable(pageSizeProp) ? clampPageSize(pageSizeProp) : pageSizeProp);
  const pageCount = getPageCount({
    total,
    pageSize
  });

  /**
   * ================================================
   *
   * Window All & clampPage
   *
   * ================================================
   *
   * 【注意】窗口空间都是闭区间
   */
  const windowAllStart = 1;
  const windowAllEnd = pageCount;
  const clampPage = React.useCallback(page => clampPageWithRange({
    page,
    minPage: windowAllStart,
    maxPage: windowAllEnd
  }), [windowAllEnd, windowAllStart]);

  /**
   * ================================================
   *
   * Current
   *
   * ================================================
   */
  const [current, setCurrent] = useControlledState(clampPage(defaultCurrent), isNonNullable(currentProp) ? clampPage(currentProp) : currentProp);
  const setCurrentWithCallback = React.useCallback(page => {
    if (disabled) {
      return;
    }
    setCurrent(page, _page => isFunction(onPageChange) && onPageChange(_page, pageSize));
  }, [disabled, onPageChange, pageSize, setCurrent]);

  /**
   * ================================================
   *
   * Windows: boundaryStart, sibling, boundaryEnd
   *
   * ================================================
   *
   * 【注意】窗口空间都是闭区间
   */

  const windowBoundaryStartEnd = boundaryCount.start;
  const windowBoundaryEndStart = pageCount + 1 - boundaryCount.end;
  const windowBoundaryEndEnd = pageCount + 1;
  const windowSiblingSize = siblingCount.start + 1 + siblingCount.end;

  /**
   * ================================================
   *
   * Pagers
   *
   * ================================================
   */
  const getPagers = React.useCallback(function (_current) {
    let {
      autoPadding = false
    } = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    const originWindowSiblingStart = _current - siblingCount.start;
    const originWindowSiblingEnd = _current + siblingCount.end;
    let windowSiblingStart = originWindowSiblingStart;
    let windowSiblingEnd = originWindowSiblingEnd;
    if (autoPadding) {
      const firstEllipsisPage = windowBoundaryStartEnd + ellipsisThreshold + siblingCount.start + 1;
      const lastEllipsisPage = windowBoundaryEndStart - ellipsisThreshold - siblingCount.end - 1;
      // 相同的时候取小的那个
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const startMax = maxBy([firstEllipsisPage - 1, firstEllipsisPage], page => getPagers(page).length);
      // 相同的时候大的那个
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const endMax = maxBy([lastEllipsisPage + 1, lastEllipsisPage], page => getPagers(page).length);
      windowSiblingStart = Math.max(Math.min(windowSiblingStart, endMax - siblingCount.start), startMax - siblingCount.start);
      windowSiblingEnd = Math.min(Math.max(windowSiblingEnd, startMax + siblingCount.end), endMax + siblingCount.end);
    }
    const pagers = [];
    for (let i = windowAllStart; i <= windowAllEnd;) {
      // 判断哪些页码要展示成 pager
      if (
      // 左边界显示为页码
      i >= 0 && i <= windowBoundaryStartEnd ||
      // 右边界显示为页码
      i >= windowBoundaryEndStart && i <= windowBoundaryEndEnd ||
      // sibling 显示为页码
      i >= windowSiblingStart && i <= windowSiblingEnd ||
      // 如果 sibling window 和 boundary window 之间距离小于 ellipsisThreshold 则处于之间的都展示为页码
      windowSiblingStart - windowBoundaryStartEnd - 1 < ellipsisThreshold && i > windowBoundaryStartEnd && i < windowSiblingStart || windowBoundaryEndStart - windowSiblingEnd - 1 < ellipsisThreshold && i > windowSiblingEnd && i < windowBoundaryEndStart ||
      // threshold = 1 时，边缘处可能会判断为显示 ellipsis
      originWindowSiblingStart <= windowSiblingStart && i >= originWindowSiblingStart && i < windowSiblingStart || originWindowSiblingEnd >= windowSiblingEnd && i <= originWindowSiblingEnd && i > windowSiblingEnd) {
        pagers.push({
          type: 'page',
          page: i
        });
        i++;
        continue;
      }
      // 判断哪些页码要展示成 jump-prev
      if (i === windowBoundaryStartEnd + 1 && windowBoundaryStartEnd < windowSiblingStart) {
        if (startEllipsisCount) {
          pagers.push({
            type: 'jump-prev'
          });
        }
        // 下面的情况则直接从 windowBoundaryStartEnd + 1 跳到 windowSiblingStart，优化性能
        if (i < windowSiblingStart) {
          i = windowSiblingStart;
          continue;
        }
        i++;
        continue;
      }
      // 判断哪些页码要展示成 jump-next
      if (i === windowSiblingEnd + 1) {
        if (endEllipsisCount) {
          pagers.push({
            type: 'jump-next'
          });
        }
        // 下面的情况则直接从 windowSiblingEnd + 1 跳到 windowBoundaryEndStart，优化性能
        if (i < windowBoundaryEndStart) {
          i = windowBoundaryEndStart;
          continue;
        }
        i++;
        continue;
      }
      i++;
    }
    return pagers;
  }, [ellipsisThreshold, endEllipsisCount, startEllipsisCount, siblingCount.end, siblingCount.start, windowAllEnd, windowBoundaryEndEnd, windowBoundaryEndStart, windowBoundaryStartEnd]);
  const pagers = [{
    type: 'prev'
  }, ...getPagers(current, {
    autoPadding: enableSiblingWindowBoundary
  }), {
    type: 'next'
  }];

  /**
   * ================================================
   *
   * Computed Values
   *
   * ================================================
   */

  const prevPage = clampPage(current - 1);
  const nextPage = clampPage(current + 1);
  const jumpPrevPage = clampPage(current - windowSiblingSize);
  const jumpNextPage = clampPage(current + windowSiblingSize);

  /**
   * ================================================
   *
   * Handlers
   *
   * ================================================
   */
  const gotoPage = React.useCallback(page => {
    setCurrentWithCallback(clampPage(page));
  }, [clampPage, setCurrentWithCallback]);
  const handleGoPrev = React.useCallback(() => {
    setCurrentWithCallback(prevPage);
  }, [prevPage, setCurrentWithCallback]);
  const handleGoNext = React.useCallback(() => {
    setCurrentWithCallback(nextPage);
  }, [nextPage, setCurrentWithCallback]);
  const handleJumpPrev = React.useCallback(() => {
    if (ellipsisReadOnly) {
      return;
    }
    setCurrentWithCallback(jumpPrevPage);
  }, [ellipsisReadOnly, jumpPrevPage, setCurrentWithCallback]);
  const handleJumpNext = React.useCallback(() => {
    if (ellipsisReadOnly) {
      return;
    }
    setCurrentWithCallback(jumpNextPage);
  }, [ellipsisReadOnly, jumpNextPage, setCurrentWithCallback]);
  const handlePageSizeChange = React.useCallback(_pageSize => {
    if (disabled) {
      return;
    }
    const newCurrent = clampPageWithRange({
      page: current,
      minPage: windowAllStart,
      maxPage: getPageCount({
        total,
        pageSize: _pageSize
      })
    });
    setCurrent(newCurrent);
    setPageSize(_pageSize);
    isFunction(onPageChange) && onPageChange(newCurrent, _pageSize);
    isFunction(onPageSizeChange) && onPageSizeChange(newCurrent, _pageSize);
  }, [current, disabled, onPageChange, onPageSizeChange, setCurrent, setPageSize, total]);
  return {
    total,
    current,
    pageSize,
    pagers,
    prevDisabled: current <= 1,
    nextDisabled: current >= pageCount,
    prevPage,
    nextPage,
    jumpPrevPage,
    jumpNextPage,
    ellipsisReadOnly,
    gotoPage,
    handleGoPrev,
    handleGoNext,
    handleJumpPrev,
    handleJumpNext,
    handlePageSizeChange
  };
};
