// @byted-service/fetch-protocol-faas/lib/consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.idcToDomainSuffix = exports.regionToDomainSuffix = exports.gatewayPSM = exports.warnings = void 0;
const tslib_1 = require("tslib");
const cn_json_1 = require("./config/cn.json");
const us_json_1 = require("./config/us.json");
const boe_json_1 = require("./config/boe.json");
exports.warnings = tslib_1.__importStar(require("./warnings"));
/**
 * https://bytedance.feishu.cn/docs/doccnSI5ppy33qV2QOtnAiw3WBb
 */
exports.gatewayPSM = 'toutiao.bytefaas.gateway';
/**
 * https://bytedance.feishu.cn/wiki/wikcnIDqFcfRLvINChAn8jIvv0b#
 */
exports.regionToDomainSuffix = {
    CN: cn_json_1.regionToDomainSuffix.CN,
    MALIVA: us_json_1.regionToDomainSuffix.MALIVA,
    ALISG: us_json_1.regionToDomainSuffix.ALISG,
    BOE: boe_json_1.regionToDomainSuffix.BOE,
};
/**
 * https://bytedance.feishu.cn/wiki/wikcnIDqFcfRLvINChAn8jIvv0b#
 */
exports.idcToDomainSuffix = {
    lf: cn_json_1.idcToDomainSuffix.lf,
    hl: cn_json_1.idcToDomainSuffix.hl,
    lq: cn_json_1.idcToDomainSuffix.lq,
    maliva: us_json_1.idcToDomainSuffix.maliva,
    alisg: us_json_1.idcToDomainSuffix.alisg,
    boe: boe_json_1.idcToDomainSuffix.boe,
    boei18n: boe_json_1.idcToDomainSuffix.boei18n,
    gcp: us_json_1.idcToDomainSuffix.gcp,
    aggdsz: cn_json_1.idcToDomainSuffix.aggdsz,
    agjsnj: cn_json_1.idcToDomainSuffix.agjsnj,
    agsdqd: cn_json_1.idcToDomainSuffix.agsdqd,
    aglnsy: cn_json_1.idcToDomainSuffix.aglnsy,
    agsxxa: cn_json_1.idcToDomainSuffix.agsxxa,
    aghbwh: cn_json_1.idcToDomainSuffix.aghbwh,
    agsy: cn_json_1.idcToDomainSuffix.agsy,
};
