// @universe-design/react/src/dialog/hooks/RefDialog.tsx
import React from 'react';
import { isFunction } from '../../util';
import type { ConfirmProps } from '../Confirm';
import { Confirm } from '../Confirm';
import type { DialogProps } from '../dialog';
import type { ModalProps } from '../Modal';
import { Modal } from '../Modal';

export interface DialogIsntance<T> {
  destroy: () => void;
  update: (config: React.SetStateAction<T>) => void;
}

interface RefDialogProps<T> {
  defaultConfig?: T;
}

const createRefDialog = <
  T extends {
    visible?: boolean;
    onClose?: DialogProps['onClose'];
  },
>({
  Comp,
  componentName,
}: {
  Comp: React.ComponentType<T>;
  componentName: string;
}) => {
  const C = React.forwardRef<DialogIsntance<T>, RefDialogProps<T>>(({ defaultConfig }, ref) => {
    const [visible, setVisible] = React.useState(true);
    const [innerConfig, setInnerConfig] = React.useState<T>(defaultConfig ?? ({} as T));

    React.useImperativeHandle(ref, () => ({
      destroy: () => {
        setVisible(false);
      },
      update: newConfig => {
        setInnerConfig(
          isFunction(newConfig)
            ? newConfig
            : originConfig => ({
                ...originConfig,
                ...newConfig,
              }),
        );
      },
    }));

    return (
      <Comp
        {...({
          ...innerConfig,
          visible,
          onClose: (..._) => {
            setVisible(false);
          },
        } as T)}
      />
    );
  });

  if (__DEV__) {
    C.displayName = componentName;
  }

  return C;
};

export const RefConfirm = /* #__PURE__ */ createRefDialog<ConfirmProps>({
  Comp: Confirm,
  componentName: 'RefConfirm',
});

export const RefModal = /* #__PURE__ */ createRefDialog<ModalProps>({
  Comp: Modal,
  componentName: 'RefModal',
});
