// @universe-design/react/src/dialog/hooks/useDialog.tsx
/* eslint-disable max-nested-callbacks */
/**
 * @file useDialog
 */
import React from 'react';
import { isFunction } from '../../util';
import type { ConfirmProps } from '../Confirm';
import { withError, withInfo, withSuccess, withWarn } from '../Confirm';
import type { ModalProps } from '../Modal';
import type { DialogIsntance } from './RefDialog';
import { RefConfirm, RefModal } from './RefDialog';

interface ReactElementsRef {
  addNode: (node: React.ReactElement) => () => void;
}

/**
 * 命令式操作 react elements 的渲染
 */
const ReactElements = React.forwardRef<ReactElementsRef>((_, ref) => {
  const [elements, setElements] = React.useState<React.ReactElement[]>([]);

  const addNode = React.useCallback((element: React.ReactElement) => {
    setElements(originElements => {
      if (originElements.includes(element)) {
        return originElements;
      }
      return [...originElements, element];
    });
    return () => {
      setElements(originElements => {
        if (originElements.includes(element)) {
          return originElements.filter(ele => ele !== element);
        }
        return originElements;
      });
    };
  }, []);

  React.useImperativeHandle(
    ref,
    () => ({
      addNode,
    }),
    [addNode],
  );

  return <>{elements}</>;
});

type Task = () => void;
const EMPTY_ACTION_ARRAY: Task[] = [];
// 自增 id
let uid = 0;
/**
 * useDialog 可以生成能获取到 context 的 dialog 实例
 */
export function useDialog() {
  const reactElementsRef = React.useRef<ReactElementsRef>(null);
  // actionQueue 用于解决 elementsHolder 和 dialog 组件还未 render 时 open update destroy 方法不生效的问题，每次都把没执行成功的 task 放入队列，下次 render 之后延迟执行
  const [actionQueue, dispatch] = React.useReducer<
    React.Reducer<Task[], { type: 'enqueue'; payload: Task } | { type: 'dequeue'; payload: Task }>
  >((state, action) => {
    switch (action.type) {
      case 'enqueue': {
        return [...state, action.payload];
      }
      case 'dequeue': {
        if (!state.includes(action.payload)) {
          return state;
        }
        return state.filter(o => o !== action.payload);
      }
      default: {
        throw new Error('Action type is not valid.');
      }
    }
  }, EMPTY_ACTION_ARRAY);

  React.useEffect(() => {
    if (!actionQueue.length) {
      return;
    }
    for (const action of actionQueue) {
      dispatch({ type: 'dequeue', payload: action });
      action();
    }
  }, [actionQueue]);

  const openDialog = React.useCallback(
    <
      T extends {
        onAfterClose?: () => void;
      },
    >(
      { onAfterClose, ...props }: T,
      RefComp: React.ForwardRefExoticComponent<{ defaultConfig?: T } & React.RefAttributes<DialogIsntance<T>>>,
    ): DialogIsntance<T> => {
      uid += 1;
      const dialogRef = React.createRef<DialogIsntance<T>>();
      const addNode = () => {
        const unmount = reactElementsRef.current?.addNode(
          <RefComp
            key={`dialog-${uid}`}
            ref={dialogRef}
            defaultConfig={
              {
                ...props,
                // after close unmount 节点
                onAfterClose: () => {
                  isFunction(onAfterClose) && onAfterClose();
                  unmount?.();
                },
              } as T
            }
          />,
        );
      };

      if (reactElementsRef.current) {
        addNode();
      } else {
        dispatch({ type: 'enqueue', payload: addNode });
      }

      const update = (config: React.SetStateAction<T>) => {
        if (dialogRef.current) {
          dialogRef.current.update(config);
        } else {
          dispatch({ type: 'enqueue', payload: () => update(config) });
        }
      };

      const destroy = () => {
        if (dialogRef.current) {
          dialogRef.current.destroy();
        } else {
          dispatch({ type: 'enqueue', payload: () => destroy() });
        }
      };

      return {
        destroy,
        update,
      };
    },
    [],
  );

  const dialogInstance = React.useMemo(
    () => ({
      confirm: (props: ConfirmProps) => openDialog(props, RefConfirm),
      info: (props: ConfirmProps) => openDialog(withInfo(props), RefConfirm),
      success: (props: ConfirmProps) => openDialog(withSuccess(props), RefConfirm),
      error: (props: ConfirmProps) => openDialog(withError(props), RefConfirm),
      warn: (props: ConfirmProps) => openDialog(withWarn(props), RefConfirm),
      open: (props: ModalProps) => openDialog(props, RefModal),
    }),
    [openDialog],
  );

  const reactElements = React.useMemo(() => <ReactElements ref={reactElementsRef} />, []);

  return [dialogInstance, reactElements] as const;
}
