// @universe-design/react/lib/divider/sx.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.sxWithTextDashed = exports.sxWithText = exports.sxTitle = exports.sxRight = exports.sxLeft = exports.sxDivider = exports.sxDirection = exports.sxDashed = void 0;
var _u1fdqv2r = _interopRequireDefault(require("./../.stylex/u1fdqv2r"));
var _u2fkmih = _interopRequireDefault(require("./../.stylex/u2fkmih"));
var _u6qvdto = _interopRequireDefault(require("./../.stylex/u6qvdto"));
var _u19tdsxb = _interopRequireDefault(require("./../.stylex/u19tdsxb"));
var _u1p85sbs = _interopRequireDefault(require("./../.stylex/u1p85sbs"));
var _ultm4w = _interopRequireDefault(require("./../.stylex/ultm4w8"));
var _ulphrlj = _interopRequireDefault(require("./../.stylex/ulphrlj"));
var _u16kelhn = _interopRequireDefault(require("./../.stylex/u16kelhn"));
var _u5jxv = _interopRequireDefault(require("./../.stylex/u5jxv71"));
var _u1jf9bby = _interopRequireDefault(require("./../.stylex/u1jf9bby"));
var _u8wgnjj = _interopRequireDefault(require("./../.stylex/u8wgnjj"));
var _uq5x3qs = _interopRequireDefault(require("./../.stylex/uq5x3qs"));
var _u57ipeq = _interopRequireDefault(require("./../.stylex/u57ipeq"));
var _ujgl7ix = _interopRequireDefault(require("./../.stylex/ujgl7ix"));
var _u1o4axus = _interopRequireDefault(require("./../.stylex/u1o4axus"));
var _u5msrnt = _interopRequireDefault(require("./../.stylex/u5msrnt"));
var _u1uvgrag = _interopRequireDefault(require("./../.stylex/u1uvgrag"));
var _uvn5txf = _interopRequireDefault(require("./../.stylex/uvn5txf"));
var _ur = _interopRequireDefault(require("./../.stylex/ur61548"));
var _u1az68f = _interopRequireDefault(require("./../.stylex/u1az68f"));
var _ukchsq = _interopRequireDefault(require("./../.stylex/ukchsq7"));
var _uxrqna = _interopRequireDefault(require("./../.stylex/uxrqna1"));
var _u13d3rh = _interopRequireDefault(require("./../.stylex/u13d3rh8"));
var _u17002dn = _interopRequireDefault(require("./../.stylex/u17002dn"));
var _u1nqr7ng = _interopRequireDefault(require("./../.stylex/u1nqr7ng"));
var _u1purjcd = _interopRequireDefault(require("./../.stylex/u1purjcd"));
var _u1aas5tn = _interopRequireDefault(require("./../.stylex/u1aas5tn"));
var _u2o5q4q = _interopRequireDefault(require("./../.stylex/u2o5q4q"));
var _uibgean = _interopRequireDefault(require("./../.stylex/uibgean"));
var _u3a4h1v = _interopRequireDefault(require("./../.stylex/u3a4h1v"));
var _u16hknla = _interopRequireDefault(require("./../.stylex/u16hknla"));
var _u1o7gc4c = _interopRequireDefault(require("./../.stylex/u1o7gc4c"));
var _u1r3828z = _interopRequireDefault(require("./../.stylex/u1r3828z"));
var _uugkpzq = _interopRequireDefault(require("./../.stylex/uugkpzq"));
var _u1ust6hi = _interopRequireDefault(require("./../.stylex/u1ust6hi"));
var _u1e5k9l = _interopRequireDefault(require("./../.stylex/u1e5k9l6"));
var _umyyz0k = _interopRequireDefault(require("./../.stylex/umyyz0k"));
var _u18cctnp = _interopRequireDefault(require("./../.stylex/u18cctnp"));
var _ujxjq4t = _interopRequireDefault(require("./../.stylex/ujxjq4t"));
var _u97qbzj = _interopRequireDefault(require("./../.stylex/u97qbzj"));
var _ulvenvz = _interopRequireDefault(require("./../.stylex/ulvenvz"));
var _u1i5fk3d = _interopRequireDefault(require("./../.stylex/u1i5fk3d"));
var _u4w7uz = _interopRequireDefault(require("./../.stylex/u4w7uz9"));
var _u1hpv8lm = _interopRequireDefault(require("./../.stylex/u1hpv8lm"));
var _u3fxpxy = _interopRequireDefault(require("./../.stylex/u3fxpxy"));
var _uwrea = _interopRequireDefault(require("./../.stylex/uwrea88"));
var _u7kv12x = _interopRequireDefault(require("./../.stylex/u7kv12x"));
var _udzhurv = _interopRequireDefault(require("./../.stylex/udzhurv"));
var _uzpagyr = _interopRequireDefault(require("./../.stylex/uzpagyr"));
var _u1lymrsk = _interopRequireDefault(require("./../.stylex/u1lymrsk"));
var _u1rg8mk = _interopRequireDefault(require("./../.stylex/u1rg8mk8"));
var _uabpw5u = _interopRequireDefault(require("./../.stylex/uabpw5u"));
var _u7grb = _interopRequireDefault(require("./../.stylex/u7grb02"));
var _u1cpcpb = _interopRequireDefault(require("./../.stylex/u1cpcpb5"));
var _us9j13g = _interopRequireDefault(require("./../.stylex/us9j13g"));
var _uzthjd = _interopRequireDefault(require("./../.stylex/uzthjd1"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
//.u1g0x7ka{padding-top:0;}
//.ude2wo7{padding-right:1em;}
//.u1qch9an{padding-bottom:0;}
//.uyan362{padding-left:1em;}
//.u1ewtqcl{box-sizing:border-box;}
//.upnsfar::before{border-top-style:dashed;}
//.u1k6lsn2::after{border-top-style:dashed;}
//.untkm7k{border-top-style:dashed;}
//.u10pi13n{position:relative;}
//.uptdccv{top:-.06em;}
//.u14t3ns0{display:inline-block;}
//.u1sdqz7s{height:.9em;}
//.u1hu3pq6{margin-top:0;}
//.u1phki43{margin-right:8px;}
//.u19f4twv{margin-bottom:0;}
//.uf9s3yw{margin-left:8px;}
//.umrv4ls{vertical-align:middle;}
//.uinvdd3{border-left-width:1px;}
//.ujik90z{border-left-style:solid;}
//.uzxpwcz{border-left-color:@line-divider-default;}
//.u22iagw{display:flex;}
//.u1hgrdrl{clear:both;}
//.uly5x3f{width:100%;}
//.u11qra4b{min-width:100%;}
//.uyycim2{margin-top:24px;}
//.u11qmguv{margin-right:0;}
//.uvueend{margin-bottom:24px;}
//.u1tyq0we{margin-left:0;}
//.umd54rt::before{width:95%;}
//.utqlky5::after{width:5%;}
//.u1q25ane::before{width:5%;}
//.uiefn6m::after{width:95%;}
//.u1pit55b{border-top-width:0px;}
//.uz5stix{white-space:nowrap;}
//.u17mccla{text-align:center;}
//.u1s5rke0::before{position:relative;}
//.u15lm7zd::before{top:50%;}
//.ulv9ixv::before{width:50%;}
//.u1fjh9ub::before{border-top-width:1px;}
//.u163fonl::before{border-top-style:solid;}
//.u1v9ev1j::before{border-top-color:inherit;}
//.u22404b::before{transform:translateY(50%);}
//.uyl8oag::before{box-sizing:border-box;}
//.u1wl9k8s::before{content:"";}
//.ut8057j::after{position:relative;}
//.ummikit::after{top:50%;}
//.u1slxich::after{width:50%;}
//.ua5nh5::after{border-top-width:1px;}
//.u8rth92::after{border-top-style:solid;}
//.uhtl3ys::after{border-top-color:inherit;}
//.unbjpgb::after{transform:translateY(50%);}
//.u1gw3sf2::after{box-sizing:border-box;}
//.u13zj6fq::after{content:"";}
//.u5ogflp{border-top-width:1px;}
//.uzkkow9{border-top-style:solid;}
//.u1yiz13w{border-top-color:@line-divider-default;}
const sxDivider = exports.sxDivider = {
  "props": {
    "B4j52fo": _u1cpcpb.default,
    "icvyot": _us9j13g.default,
    "g2u3we": _uzthjd.default
  }
};
const sxWithText = exports.sxWithText = {
  "props": {
    "B4j52fo": _u1r3828z.default,
    "Huce71": _uugkpzq.default,
    "fsow6f": _u1ust6hi.default,
    "Brfgrao": _u1e5k9l.default,
    "Bciustq": _umyyz0k.default,
    "Ccq8qp": _u18cctnp.default,
    "B0qfbqy": _ujxjq4t.default,
    "Bdkvgpv": _u97qbzj.default,
    "Bq4z7u6": _ulvenvz.default,
    "Hv9wc6": _u1i5fk3d.default,
    "Bcvre1j": _u4w7uz.default,
    "Ftih45": _u1hpv8lm.default,
    "E3zdtr": _u3fxpxy.default,
    "bn5sak": _uwrea.default,
    "a2br6o": _u7kv12x.default,
    "xrcqlc": _udzhurv.default,
    "Bm2nyyq": _uzpagyr.default,
    "B0n5ga8": _u1lymrsk.default,
    "b1kco5": _u1rg8mk.default,
    "li1rpt": _uabpw5u.default,
    "Bsft5z2": _u7grb.default
  }
};
const sxLeft = exports.sxLeft = {
  "props": {
    "Bciustq": _umyyz0k.default,
    "Ccq8qp": _u16hknla.default,
    "bn5sak": _uwrea.default,
    "a2br6o": _u1o7gc4c.default
  }
};
const sxRight = exports.sxRight = {
  "props": {
    "Bciustq": _umyyz0k.default,
    "Ccq8qp": _uibgean.default,
    "bn5sak": _uwrea.default,
    "a2br6o": _u3a4h1v.default
  }
};
const sxHorizontal = {
  "props": {
    "mc9l5x": _ukchsq.default,
    "B2qlsxq": _uxrqna.default,
    "a9b677": _u13d3rh.default,
    "Bf4jedk": _u17002dn.default,
    "B6of3ja": _u1nqr7ng.default,
    "t21cq0": _u1purjcd.default,
    "jrapky": _u1aas5tn.default,
    "Frg6f3": _u2o5q4q.default
  }
};
const sxVertical = {
  "props": {
    "qhf8xq": _u5jxv.default,
    "Bhzewxz": _u1jf9bby.default,
    "mc9l5x": _u8wgnjj.default,
    "Bqenvij": _uq5x3qs.default,
    "B6of3ja": _u57ipeq.default,
    "t21cq0": _ujgl7ix.default,
    "jrapky": _u1o4axus.default,
    "Frg6f3": _u5msrnt.default,
    "ha4doy": _u1uvgrag.default,
    "B4j52fo": _u1r3828z.default,
    "ibv6hh": _uvn5txf.default,
    "wvpqe5": _ur.default,
    "zhjwy3": _u1az68f.default
  }
};
const sxDirection = type => {
  if (type === 'horizontal') {
    return sxHorizontal;
  } else {
    if (type === 'vertical') {
      return sxVertical;
    }
  }
};
exports.sxDirection = sxDirection;
const sxDashed = exports.sxDashed = {
  "props": {
    "icvyot": _u16kelhn.default
  }
};
const sxWithTextDashed = exports.sxWithTextDashed = {
  "props": {
    "Bdkvgpv": _ultm4w.default,
    "Bm2nyyq": _ulphrlj.default
  }
};
const sxTitle = exports.sxTitle = {
  "props": {
    "mc9l5x": _u8wgnjj.default,
    "z8tnut": _u1fdqv2r.default,
    "z189sj": _u2fkmih.default,
    "Byoj8tv": _u6qvdto.default,
    "uwmqm3": _u19tdsxb.default,
    "B7ck84d": _u1p85sbs.default
  }
};
