// @byted-service/zti-helper/lib/gen/workload_pb.js
// source: workload.proto
/**
 * @fileoverview
 * @enhanceable
 * @suppress {missingRequire} reports error on implicit type usages.
 * @suppress {messageConventions} JS Compiler reports an error if a variable or
 *     field starts with 'MSG_' and isn't a translatable message.
 * @public
 */
// GENERATED CODE -- DO NOT EDIT!
/* eslint-disable */
// @ts-nocheck

var jspb = require('google-protobuf');
var goog = jspb;
var global = Function('return this')();

var google_protobuf_struct_pb = require('google-protobuf/google/protobuf/struct_pb.js');
goog.object.extend(proto, google_protobuf_struct_pb);
goog.exportSymbol('proto.JWTBundlesRequest', null, global);
goog.exportSymbol('proto.JWTBundlesResponse', null, global);
goog.exportSymbol('proto.JWTSVID', null, global);
goog.exportSymbol('proto.JWTSVIDRequest', null, global);
goog.exportSymbol('proto.JWTSVIDResponse', null, global);
goog.exportSymbol('proto.ValidateJWTSVIDRequest', null, global);
goog.exportSymbol('proto.ValidateJWTSVIDResponse', null, global);
goog.exportSymbol('proto.X509SVID', null, global);
goog.exportSymbol('proto.X509SVIDRequest', null, global);
goog.exportSymbol('proto.X509SVIDResponse', null, global);
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.X509SVIDRequest = function (opt_data) {
    jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.X509SVIDRequest, jspb.Message);
if (goog.DEBUG && !COMPILED) {
    /**
     * @public
     * @override
     */
    proto.X509SVIDRequest.displayName = 'proto.X509SVIDRequest';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.X509SVIDResponse = function (opt_data) {
    jspb.Message.initialize(
        this,
        opt_data,
        0,
        -1,
        proto.X509SVIDResponse.repeatedFields_,
        null
    );
};
goog.inherits(proto.X509SVIDResponse, jspb.Message);
if (goog.DEBUG && !COMPILED) {
    /**
     * @public
     * @override
     */
    proto.X509SVIDResponse.displayName = 'proto.X509SVIDResponse';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.X509SVID = function (opt_data) {
    jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.X509SVID, jspb.Message);
if (goog.DEBUG && !COMPILED) {
    /**
     * @public
     * @override
     */
    proto.X509SVID.displayName = 'proto.X509SVID';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.JWTSVID = function (opt_data) {
    jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.JWTSVID, jspb.Message);
if (goog.DEBUG && !COMPILED) {
    /**
     * @public
     * @override
     */
    proto.JWTSVID.displayName = 'proto.JWTSVID';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.JWTSVIDRequest = function (opt_data) {
    jspb.Message.initialize(
        this,
        opt_data,
        0,
        -1,
        proto.JWTSVIDRequest.repeatedFields_,
        null
    );
};
goog.inherits(proto.JWTSVIDRequest, jspb.Message);
if (goog.DEBUG && !COMPILED) {
    /**
     * @public
     * @override
     */
    proto.JWTSVIDRequest.displayName = 'proto.JWTSVIDRequest';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.JWTSVIDResponse = function (opt_data) {
    jspb.Message.initialize(
        this,
        opt_data,
        0,
        -1,
        proto.JWTSVIDResponse.repeatedFields_,
        null
    );
};
goog.inherits(proto.JWTSVIDResponse, jspb.Message);
if (goog.DEBUG && !COMPILED) {
    /**
     * @public
     * @override
     */
    proto.JWTSVIDResponse.displayName = 'proto.JWTSVIDResponse';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.JWTBundlesRequest = function (opt_data) {
    jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.JWTBundlesRequest, jspb.Message);
if (goog.DEBUG && !COMPILED) {
    /**
     * @public
     * @override
     */
    proto.JWTBundlesRequest.displayName = 'proto.JWTBundlesRequest';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.JWTBundlesResponse = function (opt_data) {
    jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.JWTBundlesResponse, jspb.Message);
if (goog.DEBUG && !COMPILED) {
    /**
     * @public
     * @override
     */
    proto.JWTBundlesResponse.displayName = 'proto.JWTBundlesResponse';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.ValidateJWTSVIDRequest = function (opt_data) {
    jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.ValidateJWTSVIDRequest, jspb.Message);
if (goog.DEBUG && !COMPILED) {
    /**
     * @public
     * @override
     */
    proto.ValidateJWTSVIDRequest.displayName = 'proto.ValidateJWTSVIDRequest';
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.ValidateJWTSVIDResponse = function (opt_data) {
    jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.ValidateJWTSVIDResponse, jspb.Message);
if (goog.DEBUG && !COMPILED) {
    /**
     * @public
     * @override
     */
    proto.ValidateJWTSVIDResponse.displayName = 'proto.ValidateJWTSVIDResponse';
}

if (jspb.Message.GENERATE_TO_OBJECT) {
    /**
     * Creates an object representation of this proto.
     * Field names that are reserved in JavaScript and will be renamed to pb_name.
     * Optional fields that are not set will be set to undefined.
     * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
     * For the list of reserved names please see:
     *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
     * @param {boolean=} opt_includeInstance Deprecated. whether to include the
     *     JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @return {!Object}
     */
    proto.X509SVIDRequest.prototype.toObject = function (opt_includeInstance) {
        return proto.X509SVIDRequest.toObject(opt_includeInstance, this);
    };

    /**
     * Static version of the {@see toObject} method.
     * @param {boolean|undefined} includeInstance Deprecated. Whether to include
     *     the JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @param {!proto.X509SVIDRequest} msg The msg instance to transform.
     * @return {!Object}
     * @suppress {unusedLocalVariables} f is only used for nested messages
     */
    proto.X509SVIDRequest.toObject = function (includeInstance, msg) {
        var f,
            obj = {};

        if (includeInstance) {
            obj.$jspbMessageInstance = msg;
        }
        return obj;
    };
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.X509SVIDRequest}
 */
proto.X509SVIDRequest.deserializeBinary = function (bytes) {
    var reader = new jspb.BinaryReader(bytes);
    var msg = new proto.X509SVIDRequest();
    return proto.X509SVIDRequest.deserializeBinaryFromReader(msg, reader);
};

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.X509SVIDRequest} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.X509SVIDRequest}
 */
proto.X509SVIDRequest.deserializeBinaryFromReader = function (msg, reader) {
    while (reader.nextField()) {
        if (reader.isEndGroup()) {
            break;
        }
        var field = reader.getFieldNumber();
        switch (field) {
            default:
                reader.skipField();
                break;
        }
    }
    return msg;
};

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.X509SVIDRequest.prototype.serializeBinary = function () {
    var writer = new jspb.BinaryWriter();
    proto.X509SVIDRequest.serializeBinaryToWriter(this, writer);
    return writer.getResultBuffer();
};

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.X509SVIDRequest} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.X509SVIDRequest.serializeBinaryToWriter = function (message, writer) {
    var f = undefined;
};

/**
 * List of repeated fields within this message type.
 * @private {!Array<number>}
 * @const
 */
proto.X509SVIDResponse.repeatedFields_ = [1, 2];

if (jspb.Message.GENERATE_TO_OBJECT) {
    /**
     * Creates an object representation of this proto.
     * Field names that are reserved in JavaScript and will be renamed to pb_name.
     * Optional fields that are not set will be set to undefined.
     * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
     * For the list of reserved names please see:
     *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
     * @param {boolean=} opt_includeInstance Deprecated. whether to include the
     *     JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @return {!Object}
     */
    proto.X509SVIDResponse.prototype.toObject = function (opt_includeInstance) {
        return proto.X509SVIDResponse.toObject(opt_includeInstance, this);
    };

    /**
     * Static version of the {@see toObject} method.
     * @param {boolean|undefined} includeInstance Deprecated. Whether to include
     *     the JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @param {!proto.X509SVIDResponse} msg The msg instance to transform.
     * @return {!Object}
     * @suppress {unusedLocalVariables} f is only used for nested messages
     */
    proto.X509SVIDResponse.toObject = function (includeInstance, msg) {
        var f,
            obj = {
                svidsList: jspb.Message.toObjectList(
                    msg.getSvidsList(),
                    proto.X509SVID.toObject,
                    includeInstance
                ),
                crlList: msg.getCrlList_asB64(),
                federatedBundlesMap: (f = msg.getFederatedBundlesMap())
                    ? f.toObject(includeInstance, undefined)
                    : [],
            };

        if (includeInstance) {
            obj.$jspbMessageInstance = msg;
        }
        return obj;
    };
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.X509SVIDResponse}
 */
proto.X509SVIDResponse.deserializeBinary = function (bytes) {
    var reader = new jspb.BinaryReader(bytes);
    var msg = new proto.X509SVIDResponse();
    return proto.X509SVIDResponse.deserializeBinaryFromReader(msg, reader);
};

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.X509SVIDResponse} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.X509SVIDResponse}
 */
proto.X509SVIDResponse.deserializeBinaryFromReader = function (msg, reader) {
    while (reader.nextField()) {
        if (reader.isEndGroup()) {
            break;
        }
        var field = reader.getFieldNumber();
        switch (field) {
            case 1:
                var value = new proto.X509SVID();
                reader.readMessage(
                    value,
                    proto.X509SVID.deserializeBinaryFromReader
                );
                msg.addSvids(value);
                break;
            case 2:
                var value = /** @type {!Uint8Array} */ (reader.readBytes());
                msg.addCrl(value);
                break;
            case 3:
                var value = msg.getFederatedBundlesMap();
                reader.readMessage(value, function (message, reader) {
                    jspb.Map.deserializeBinary(
                        message,
                        reader,
                        jspb.BinaryReader.prototype.readString,
                        jspb.BinaryReader.prototype.readBytes,
                        null,
                        '',
                        ''
                    );
                });
                break;
            default:
                reader.skipField();
                break;
        }
    }
    return msg;
};

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.X509SVIDResponse.prototype.serializeBinary = function () {
    var writer = new jspb.BinaryWriter();
    proto.X509SVIDResponse.serializeBinaryToWriter(this, writer);
    return writer.getResultBuffer();
};

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.X509SVIDResponse} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.X509SVIDResponse.serializeBinaryToWriter = function (message, writer) {
    var f = undefined;
    f = message.getSvidsList();
    if (f.length > 0) {
        writer.writeRepeatedMessage(
            1,
            f,
            proto.X509SVID.serializeBinaryToWriter
        );
    }
    f = message.getCrlList_asU8();
    if (f.length > 0) {
        writer.writeRepeatedBytes(2, f);
    }
    f = message.getFederatedBundlesMap(true);
    if (f && f.getLength() > 0) {
        f.serializeBinary(
            3,
            writer,
            jspb.BinaryWriter.prototype.writeString,
            jspb.BinaryWriter.prototype.writeBytes
        );
    }
};

/**
 * repeated X509SVID svids = 1;
 * @return {!Array<!proto.X509SVID>}
 */
proto.X509SVIDResponse.prototype.getSvidsList = function () {
    return /** @type{!Array<!proto.X509SVID>} */ (
        jspb.Message.getRepeatedWrapperField(this, proto.X509SVID, 1)
    );
};

/**
 * @param {!Array<!proto.X509SVID>} value
 * @return {!proto.X509SVIDResponse} returns this
 */
proto.X509SVIDResponse.prototype.setSvidsList = function (value) {
    return jspb.Message.setRepeatedWrapperField(this, 1, value);
};

/**
 * @param {!proto.X509SVID=} opt_value
 * @param {number=} opt_index
 * @return {!proto.X509SVID}
 */
proto.X509SVIDResponse.prototype.addSvids = function (opt_value, opt_index) {
    return jspb.Message.addToRepeatedWrapperField(
        this,
        1,
        opt_value,
        proto.X509SVID,
        opt_index
    );
};

/**
 * Clears the list making it empty but non-null.
 * @return {!proto.X509SVIDResponse} returns this
 */
proto.X509SVIDResponse.prototype.clearSvidsList = function () {
    return this.setSvidsList([]);
};

/**
 * repeated bytes crl = 2;
 * @return {!(Array<!Uint8Array>|Array<string>)}
 */
proto.X509SVIDResponse.prototype.getCrlList = function () {
    return /** @type {!(Array<!Uint8Array>|Array<string>)} */ (
        jspb.Message.getRepeatedField(this, 2)
    );
};

/**
 * repeated bytes crl = 2;
 * This is a type-conversion wrapper around `getCrlList()`
 * @return {!Array<string>}
 */
proto.X509SVIDResponse.prototype.getCrlList_asB64 = function () {
    return /** @type {!Array<string>} */ (
        jspb.Message.bytesListAsB64(this.getCrlList())
    );
};

/**
 * repeated bytes crl = 2;
 * Note that Uint8Array is not supported on all browsers.
 * @see http://caniuse.com/Uint8Array
 * This is a type-conversion wrapper around `getCrlList()`
 * @return {!Array<!Uint8Array>}
 */
proto.X509SVIDResponse.prototype.getCrlList_asU8 = function () {
    return /** @type {!Array<!Uint8Array>} */ (
        jspb.Message.bytesListAsU8(this.getCrlList())
    );
};

/**
 * @param {!(Array<!Uint8Array>|Array<string>)} value
 * @return {!proto.X509SVIDResponse} returns this
 */
proto.X509SVIDResponse.prototype.setCrlList = function (value) {
    return jspb.Message.setField(this, 2, value || []);
};

/**
 * @param {!(string|Uint8Array)} value
 * @param {number=} opt_index
 * @return {!proto.X509SVIDResponse} returns this
 */
proto.X509SVIDResponse.prototype.addCrl = function (value, opt_index) {
    return jspb.Message.addToRepeatedField(this, 2, value, opt_index);
};

/**
 * Clears the list making it empty but non-null.
 * @return {!proto.X509SVIDResponse} returns this
 */
proto.X509SVIDResponse.prototype.clearCrlList = function () {
    return this.setCrlList([]);
};

/**
 * map<string, bytes> federated_bundles = 3;
 * @param {boolean=} opt_noLazyCreate Do not create the map if
 * empty, instead returning `undefined`
 * @return {!jspb.Map<string,!(string|Uint8Array)>}
 */
proto.X509SVIDResponse.prototype.getFederatedBundlesMap = function (
    opt_noLazyCreate
) {
    return /** @type {!jspb.Map<string,!(string|Uint8Array)>} */ (
        jspb.Message.getMapField(this, 3, opt_noLazyCreate, null)
    );
};

/**
 * Clears values from the map. The map will be non-null.
 * @return {!proto.X509SVIDResponse} returns this
 */
proto.X509SVIDResponse.prototype.clearFederatedBundlesMap = function () {
    this.getFederatedBundlesMap().clear();
    return this;
};

if (jspb.Message.GENERATE_TO_OBJECT) {
    /**
     * Creates an object representation of this proto.
     * Field names that are reserved in JavaScript and will be renamed to pb_name.
     * Optional fields that are not set will be set to undefined.
     * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
     * For the list of reserved names please see:
     *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
     * @param {boolean=} opt_includeInstance Deprecated. whether to include the
     *     JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @return {!Object}
     */
    proto.X509SVID.prototype.toObject = function (opt_includeInstance) {
        return proto.X509SVID.toObject(opt_includeInstance, this);
    };

    /**
     * Static version of the {@see toObject} method.
     * @param {boolean|undefined} includeInstance Deprecated. Whether to include
     *     the JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @param {!proto.X509SVID} msg The msg instance to transform.
     * @return {!Object}
     * @suppress {unusedLocalVariables} f is only used for nested messages
     */
    proto.X509SVID.toObject = function (includeInstance, msg) {
        var f,
            obj = {
                spiffeId: jspb.Message.getFieldWithDefault(msg, 1, ''),
                x509Svid: msg.getX509Svid_asB64(),
                x509SvidKey: msg.getX509SvidKey_asB64(),
                bundle: msg.getBundle_asB64(),
            };

        if (includeInstance) {
            obj.$jspbMessageInstance = msg;
        }
        return obj;
    };
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.X509SVID}
 */
proto.X509SVID.deserializeBinary = function (bytes) {
    var reader = new jspb.BinaryReader(bytes);
    var msg = new proto.X509SVID();
    return proto.X509SVID.deserializeBinaryFromReader(msg, reader);
};

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.X509SVID} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.X509SVID}
 */
proto.X509SVID.deserializeBinaryFromReader = function (msg, reader) {
    while (reader.nextField()) {
        if (reader.isEndGroup()) {
            break;
        }
        var field = reader.getFieldNumber();
        switch (field) {
            case 1:
                var value = /** @type {string} */ (reader.readString());
                msg.setSpiffeId(value);
                break;
            case 2:
                var value = /** @type {!Uint8Array} */ (reader.readBytes());
                msg.setX509Svid(value);
                break;
            case 3:
                var value = /** @type {!Uint8Array} */ (reader.readBytes());
                msg.setX509SvidKey(value);
                break;
            case 4:
                var value = /** @type {!Uint8Array} */ (reader.readBytes());
                msg.setBundle(value);
                break;
            default:
                reader.skipField();
                break;
        }
    }
    return msg;
};

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.X509SVID.prototype.serializeBinary = function () {
    var writer = new jspb.BinaryWriter();
    proto.X509SVID.serializeBinaryToWriter(this, writer);
    return writer.getResultBuffer();
};

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.X509SVID} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.X509SVID.serializeBinaryToWriter = function (message, writer) {
    var f = undefined;
    f = message.getSpiffeId();
    if (f.length > 0) {
        writer.writeString(1, f);
    }
    f = message.getX509Svid_asU8();
    if (f.length > 0) {
        writer.writeBytes(2, f);
    }
    f = message.getX509SvidKey_asU8();
    if (f.length > 0) {
        writer.writeBytes(3, f);
    }
    f = message.getBundle_asU8();
    if (f.length > 0) {
        writer.writeBytes(4, f);
    }
};

/**
 * optional string spiffe_id = 1;
 * @return {string}
 */
proto.X509SVID.prototype.getSpiffeId = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 1, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.X509SVID} returns this
 */
proto.X509SVID.prototype.setSpiffeId = function (value) {
    return jspb.Message.setProto3StringField(this, 1, value);
};

/**
 * optional bytes x509_svid = 2;
 * @return {!(string|Uint8Array)}
 */
proto.X509SVID.prototype.getX509Svid = function () {
    return /** @type {!(string|Uint8Array)} */ (
        jspb.Message.getFieldWithDefault(this, 2, '')
    );
};

/**
 * optional bytes x509_svid = 2;
 * This is a type-conversion wrapper around `getX509Svid()`
 * @return {string}
 */
proto.X509SVID.prototype.getX509Svid_asB64 = function () {
    return /** @type {string} */ (jspb.Message.bytesAsB64(this.getX509Svid()));
};

/**
 * optional bytes x509_svid = 2;
 * Note that Uint8Array is not supported on all browsers.
 * @see http://caniuse.com/Uint8Array
 * This is a type-conversion wrapper around `getX509Svid()`
 * @return {!Uint8Array}
 */
proto.X509SVID.prototype.getX509Svid_asU8 = function () {
    return /** @type {!Uint8Array} */ (
        jspb.Message.bytesAsU8(this.getX509Svid())
    );
};

/**
 * @param {!(string|Uint8Array)} value
 * @return {!proto.X509SVID} returns this
 */
proto.X509SVID.prototype.setX509Svid = function (value) {
    return jspb.Message.setProto3BytesField(this, 2, value);
};

/**
 * optional bytes x509_svid_key = 3;
 * @return {!(string|Uint8Array)}
 */
proto.X509SVID.prototype.getX509SvidKey = function () {
    return /** @type {!(string|Uint8Array)} */ (
        jspb.Message.getFieldWithDefault(this, 3, '')
    );
};

/**
 * optional bytes x509_svid_key = 3;
 * This is a type-conversion wrapper around `getX509SvidKey()`
 * @return {string}
 */
proto.X509SVID.prototype.getX509SvidKey_asB64 = function () {
    return /** @type {string} */ (
        jspb.Message.bytesAsB64(this.getX509SvidKey())
    );
};

/**
 * optional bytes x509_svid_key = 3;
 * Note that Uint8Array is not supported on all browsers.
 * @see http://caniuse.com/Uint8Array
 * This is a type-conversion wrapper around `getX509SvidKey()`
 * @return {!Uint8Array}
 */
proto.X509SVID.prototype.getX509SvidKey_asU8 = function () {
    return /** @type {!Uint8Array} */ (
        jspb.Message.bytesAsU8(this.getX509SvidKey())
    );
};

/**
 * @param {!(string|Uint8Array)} value
 * @return {!proto.X509SVID} returns this
 */
proto.X509SVID.prototype.setX509SvidKey = function (value) {
    return jspb.Message.setProto3BytesField(this, 3, value);
};

/**
 * optional bytes bundle = 4;
 * @return {!(string|Uint8Array)}
 */
proto.X509SVID.prototype.getBundle = function () {
    return /** @type {!(string|Uint8Array)} */ (
        jspb.Message.getFieldWithDefault(this, 4, '')
    );
};

/**
 * optional bytes bundle = 4;
 * This is a type-conversion wrapper around `getBundle()`
 * @return {string}
 */
proto.X509SVID.prototype.getBundle_asB64 = function () {
    return /** @type {string} */ (jspb.Message.bytesAsB64(this.getBundle()));
};

/**
 * optional bytes bundle = 4;
 * Note that Uint8Array is not supported on all browsers.
 * @see http://caniuse.com/Uint8Array
 * This is a type-conversion wrapper around `getBundle()`
 * @return {!Uint8Array}
 */
proto.X509SVID.prototype.getBundle_asU8 = function () {
    return /** @type {!Uint8Array} */ (
        jspb.Message.bytesAsU8(this.getBundle())
    );
};

/**
 * @param {!(string|Uint8Array)} value
 * @return {!proto.X509SVID} returns this
 */
proto.X509SVID.prototype.setBundle = function (value) {
    return jspb.Message.setProto3BytesField(this, 4, value);
};

if (jspb.Message.GENERATE_TO_OBJECT) {
    /**
     * Creates an object representation of this proto.
     * Field names that are reserved in JavaScript and will be renamed to pb_name.
     * Optional fields that are not set will be set to undefined.
     * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
     * For the list of reserved names please see:
     *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
     * @param {boolean=} opt_includeInstance Deprecated. whether to include the
     *     JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @return {!Object}
     */
    proto.JWTSVID.prototype.toObject = function (opt_includeInstance) {
        return proto.JWTSVID.toObject(opt_includeInstance, this);
    };

    /**
     * Static version of the {@see toObject} method.
     * @param {boolean|undefined} includeInstance Deprecated. Whether to include
     *     the JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @param {!proto.JWTSVID} msg The msg instance to transform.
     * @return {!Object}
     * @suppress {unusedLocalVariables} f is only used for nested messages
     */
    proto.JWTSVID.toObject = function (includeInstance, msg) {
        var f,
            obj = {
                spiffeId: jspb.Message.getFieldWithDefault(msg, 1, ''),
                svid: jspb.Message.getFieldWithDefault(msg, 2, ''),
            };

        if (includeInstance) {
            obj.$jspbMessageInstance = msg;
        }
        return obj;
    };
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.JWTSVID}
 */
proto.JWTSVID.deserializeBinary = function (bytes) {
    var reader = new jspb.BinaryReader(bytes);
    var msg = new proto.JWTSVID();
    return proto.JWTSVID.deserializeBinaryFromReader(msg, reader);
};

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.JWTSVID} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.JWTSVID}
 */
proto.JWTSVID.deserializeBinaryFromReader = function (msg, reader) {
    while (reader.nextField()) {
        if (reader.isEndGroup()) {
            break;
        }
        var field = reader.getFieldNumber();
        switch (field) {
            case 1:
                var value = /** @type {string} */ (reader.readString());
                msg.setSpiffeId(value);
                break;
            case 2:
                var value = /** @type {string} */ (reader.readString());
                msg.setSvid(value);
                break;
            default:
                reader.skipField();
                break;
        }
    }
    return msg;
};

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.JWTSVID.prototype.serializeBinary = function () {
    var writer = new jspb.BinaryWriter();
    proto.JWTSVID.serializeBinaryToWriter(this, writer);
    return writer.getResultBuffer();
};

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.JWTSVID} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.JWTSVID.serializeBinaryToWriter = function (message, writer) {
    var f = undefined;
    f = message.getSpiffeId();
    if (f.length > 0) {
        writer.writeString(1, f);
    }
    f = message.getSvid();
    if (f.length > 0) {
        writer.writeString(2, f);
    }
};

/**
 * optional string spiffe_id = 1;
 * @return {string}
 */
proto.JWTSVID.prototype.getSpiffeId = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 1, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.JWTSVID} returns this
 */
proto.JWTSVID.prototype.setSpiffeId = function (value) {
    return jspb.Message.setProto3StringField(this, 1, value);
};

/**
 * optional string svid = 2;
 * @return {string}
 */
proto.JWTSVID.prototype.getSvid = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 2, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.JWTSVID} returns this
 */
proto.JWTSVID.prototype.setSvid = function (value) {
    return jspb.Message.setProto3StringField(this, 2, value);
};

/**
 * List of repeated fields within this message type.
 * @private {!Array<number>}
 * @const
 */
proto.JWTSVIDRequest.repeatedFields_ = [1];

if (jspb.Message.GENERATE_TO_OBJECT) {
    /**
     * Creates an object representation of this proto.
     * Field names that are reserved in JavaScript and will be renamed to pb_name.
     * Optional fields that are not set will be set to undefined.
     * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
     * For the list of reserved names please see:
     *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
     * @param {boolean=} opt_includeInstance Deprecated. whether to include the
     *     JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @return {!Object}
     */
    proto.JWTSVIDRequest.prototype.toObject = function (opt_includeInstance) {
        return proto.JWTSVIDRequest.toObject(opt_includeInstance, this);
    };

    /**
     * Static version of the {@see toObject} method.
     * @param {boolean|undefined} includeInstance Deprecated. Whether to include
     *     the JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @param {!proto.JWTSVIDRequest} msg The msg instance to transform.
     * @return {!Object}
     * @suppress {unusedLocalVariables} f is only used for nested messages
     */
    proto.JWTSVIDRequest.toObject = function (includeInstance, msg) {
        var f,
            obj = {
                audienceList:
                    (f = jspb.Message.getRepeatedField(msg, 1)) == null
                        ? undefined
                        : f,
                spiffeId: jspb.Message.getFieldWithDefault(msg, 2, ''),
            };

        if (includeInstance) {
            obj.$jspbMessageInstance = msg;
        }
        return obj;
    };
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.JWTSVIDRequest}
 */
proto.JWTSVIDRequest.deserializeBinary = function (bytes) {
    var reader = new jspb.BinaryReader(bytes);
    var msg = new proto.JWTSVIDRequest();
    return proto.JWTSVIDRequest.deserializeBinaryFromReader(msg, reader);
};

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.JWTSVIDRequest} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.JWTSVIDRequest}
 */
proto.JWTSVIDRequest.deserializeBinaryFromReader = function (msg, reader) {
    while (reader.nextField()) {
        if (reader.isEndGroup()) {
            break;
        }
        var field = reader.getFieldNumber();
        switch (field) {
            case 1:
                var value = /** @type {string} */ (reader.readString());
                msg.addAudience(value);
                break;
            case 2:
                var value = /** @type {string} */ (reader.readString());
                msg.setSpiffeId(value);
                break;
            default:
                reader.skipField();
                break;
        }
    }
    return msg;
};

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.JWTSVIDRequest.prototype.serializeBinary = function () {
    var writer = new jspb.BinaryWriter();
    proto.JWTSVIDRequest.serializeBinaryToWriter(this, writer);
    return writer.getResultBuffer();
};

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.JWTSVIDRequest} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.JWTSVIDRequest.serializeBinaryToWriter = function (message, writer) {
    var f = undefined;
    f = message.getAudienceList();
    if (f.length > 0) {
        writer.writeRepeatedString(1, f);
    }
    f = message.getSpiffeId();
    if (f.length > 0) {
        writer.writeString(2, f);
    }
};

/**
 * repeated string audience = 1;
 * @return {!Array<string>}
 */
proto.JWTSVIDRequest.prototype.getAudienceList = function () {
    return /** @type {!Array<string>} */ (
        jspb.Message.getRepeatedField(this, 1)
    );
};

/**
 * @param {!Array<string>} value
 * @return {!proto.JWTSVIDRequest} returns this
 */
proto.JWTSVIDRequest.prototype.setAudienceList = function (value) {
    return jspb.Message.setField(this, 1, value || []);
};

/**
 * @param {string} value
 * @param {number=} opt_index
 * @return {!proto.JWTSVIDRequest} returns this
 */
proto.JWTSVIDRequest.prototype.addAudience = function (value, opt_index) {
    return jspb.Message.addToRepeatedField(this, 1, value, opt_index);
};

/**
 * Clears the list making it empty but non-null.
 * @return {!proto.JWTSVIDRequest} returns this
 */
proto.JWTSVIDRequest.prototype.clearAudienceList = function () {
    return this.setAudienceList([]);
};

/**
 * optional string spiffe_id = 2;
 * @return {string}
 */
proto.JWTSVIDRequest.prototype.getSpiffeId = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 2, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.JWTSVIDRequest} returns this
 */
proto.JWTSVIDRequest.prototype.setSpiffeId = function (value) {
    return jspb.Message.setProto3StringField(this, 2, value);
};

/**
 * List of repeated fields within this message type.
 * @private {!Array<number>}
 * @const
 */
proto.JWTSVIDResponse.repeatedFields_ = [1];

if (jspb.Message.GENERATE_TO_OBJECT) {
    /**
     * Creates an object representation of this proto.
     * Field names that are reserved in JavaScript and will be renamed to pb_name.
     * Optional fields that are not set will be set to undefined.
     * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
     * For the list of reserved names please see:
     *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
     * @param {boolean=} opt_includeInstance Deprecated. whether to include the
     *     JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @return {!Object}
     */
    proto.JWTSVIDResponse.prototype.toObject = function (opt_includeInstance) {
        return proto.JWTSVIDResponse.toObject(opt_includeInstance, this);
    };

    /**
     * Static version of the {@see toObject} method.
     * @param {boolean|undefined} includeInstance Deprecated. Whether to include
     *     the JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @param {!proto.JWTSVIDResponse} msg The msg instance to transform.
     * @return {!Object}
     * @suppress {unusedLocalVariables} f is only used for nested messages
     */
    proto.JWTSVIDResponse.toObject = function (includeInstance, msg) {
        var f,
            obj = {
                svidsList: jspb.Message.toObjectList(
                    msg.getSvidsList(),
                    proto.JWTSVID.toObject,
                    includeInstance
                ),
            };

        if (includeInstance) {
            obj.$jspbMessageInstance = msg;
        }
        return obj;
    };
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.JWTSVIDResponse}
 */
proto.JWTSVIDResponse.deserializeBinary = function (bytes) {
    var reader = new jspb.BinaryReader(bytes);
    var msg = new proto.JWTSVIDResponse();
    return proto.JWTSVIDResponse.deserializeBinaryFromReader(msg, reader);
};

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.JWTSVIDResponse} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.JWTSVIDResponse}
 */
proto.JWTSVIDResponse.deserializeBinaryFromReader = function (msg, reader) {
    while (reader.nextField()) {
        if (reader.isEndGroup()) {
            break;
        }
        var field = reader.getFieldNumber();
        switch (field) {
            case 1:
                var value = new proto.JWTSVID();
                reader.readMessage(
                    value,
                    proto.JWTSVID.deserializeBinaryFromReader
                );
                msg.addSvids(value);
                break;
            default:
                reader.skipField();
                break;
        }
    }
    return msg;
};

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.JWTSVIDResponse.prototype.serializeBinary = function () {
    var writer = new jspb.BinaryWriter();
    proto.JWTSVIDResponse.serializeBinaryToWriter(this, writer);
    return writer.getResultBuffer();
};

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.JWTSVIDResponse} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.JWTSVIDResponse.serializeBinaryToWriter = function (message, writer) {
    var f = undefined;
    f = message.getSvidsList();
    if (f.length > 0) {
        writer.writeRepeatedMessage(
            1,
            f,
            proto.JWTSVID.serializeBinaryToWriter
        );
    }
};

/**
 * repeated JWTSVID svids = 1;
 * @return {!Array<!proto.JWTSVID>}
 */
proto.JWTSVIDResponse.prototype.getSvidsList = function () {
    return /** @type{!Array<!proto.JWTSVID>} */ (
        jspb.Message.getRepeatedWrapperField(this, proto.JWTSVID, 1)
    );
};

/**
 * @param {!Array<!proto.JWTSVID>} value
 * @return {!proto.JWTSVIDResponse} returns this
 */
proto.JWTSVIDResponse.prototype.setSvidsList = function (value) {
    return jspb.Message.setRepeatedWrapperField(this, 1, value);
};

/**
 * @param {!proto.JWTSVID=} opt_value
 * @param {number=} opt_index
 * @return {!proto.JWTSVID}
 */
proto.JWTSVIDResponse.prototype.addSvids = function (opt_value, opt_index) {
    return jspb.Message.addToRepeatedWrapperField(
        this,
        1,
        opt_value,
        proto.JWTSVID,
        opt_index
    );
};

/**
 * Clears the list making it empty but non-null.
 * @return {!proto.JWTSVIDResponse} returns this
 */
proto.JWTSVIDResponse.prototype.clearSvidsList = function () {
    return this.setSvidsList([]);
};

if (jspb.Message.GENERATE_TO_OBJECT) {
    /**
     * Creates an object representation of this proto.
     * Field names that are reserved in JavaScript and will be renamed to pb_name.
     * Optional fields that are not set will be set to undefined.
     * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
     * For the list of reserved names please see:
     *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
     * @param {boolean=} opt_includeInstance Deprecated. whether to include the
     *     JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @return {!Object}
     */
    proto.JWTBundlesRequest.prototype.toObject = function (
        opt_includeInstance
    ) {
        return proto.JWTBundlesRequest.toObject(opt_includeInstance, this);
    };

    /**
     * Static version of the {@see toObject} method.
     * @param {boolean|undefined} includeInstance Deprecated. Whether to include
     *     the JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @param {!proto.JWTBundlesRequest} msg The msg instance to transform.
     * @return {!Object}
     * @suppress {unusedLocalVariables} f is only used for nested messages
     */
    proto.JWTBundlesRequest.toObject = function (includeInstance, msg) {
        var f,
            obj = {};

        if (includeInstance) {
            obj.$jspbMessageInstance = msg;
        }
        return obj;
    };
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.JWTBundlesRequest}
 */
proto.JWTBundlesRequest.deserializeBinary = function (bytes) {
    var reader = new jspb.BinaryReader(bytes);
    var msg = new proto.JWTBundlesRequest();
    return proto.JWTBundlesRequest.deserializeBinaryFromReader(msg, reader);
};

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.JWTBundlesRequest} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.JWTBundlesRequest}
 */
proto.JWTBundlesRequest.deserializeBinaryFromReader = function (msg, reader) {
    while (reader.nextField()) {
        if (reader.isEndGroup()) {
            break;
        }
        var field = reader.getFieldNumber();
        switch (field) {
            default:
                reader.skipField();
                break;
        }
    }
    return msg;
};

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.JWTBundlesRequest.prototype.serializeBinary = function () {
    var writer = new jspb.BinaryWriter();
    proto.JWTBundlesRequest.serializeBinaryToWriter(this, writer);
    return writer.getResultBuffer();
};

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.JWTBundlesRequest} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.JWTBundlesRequest.serializeBinaryToWriter = function (message, writer) {
    var f = undefined;
};

if (jspb.Message.GENERATE_TO_OBJECT) {
    /**
     * Creates an object representation of this proto.
     * Field names that are reserved in JavaScript and will be renamed to pb_name.
     * Optional fields that are not set will be set to undefined.
     * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
     * For the list of reserved names please see:
     *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
     * @param {boolean=} opt_includeInstance Deprecated. whether to include the
     *     JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @return {!Object}
     */
    proto.JWTBundlesResponse.prototype.toObject = function (
        opt_includeInstance
    ) {
        return proto.JWTBundlesResponse.toObject(opt_includeInstance, this);
    };

    /**
     * Static version of the {@see toObject} method.
     * @param {boolean|undefined} includeInstance Deprecated. Whether to include
     *     the JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @param {!proto.JWTBundlesResponse} msg The msg instance to transform.
     * @return {!Object}
     * @suppress {unusedLocalVariables} f is only used for nested messages
     */
    proto.JWTBundlesResponse.toObject = function (includeInstance, msg) {
        var f,
            obj = {
                bundlesMap: (f = msg.getBundlesMap())
                    ? f.toObject(includeInstance, undefined)
                    : [],
            };

        if (includeInstance) {
            obj.$jspbMessageInstance = msg;
        }
        return obj;
    };
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.JWTBundlesResponse}
 */
proto.JWTBundlesResponse.deserializeBinary = function (bytes) {
    var reader = new jspb.BinaryReader(bytes);
    var msg = new proto.JWTBundlesResponse();
    return proto.JWTBundlesResponse.deserializeBinaryFromReader(msg, reader);
};

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.JWTBundlesResponse} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.JWTBundlesResponse}
 */
proto.JWTBundlesResponse.deserializeBinaryFromReader = function (msg, reader) {
    while (reader.nextField()) {
        if (reader.isEndGroup()) {
            break;
        }
        var field = reader.getFieldNumber();
        switch (field) {
            case 1:
                var value = msg.getBundlesMap();
                reader.readMessage(value, function (message, reader) {
                    jspb.Map.deserializeBinary(
                        message,
                        reader,
                        jspb.BinaryReader.prototype.readString,
                        jspb.BinaryReader.prototype.readBytes,
                        null,
                        '',
                        ''
                    );
                });
                break;
            default:
                reader.skipField();
                break;
        }
    }
    return msg;
};

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.JWTBundlesResponse.prototype.serializeBinary = function () {
    var writer = new jspb.BinaryWriter();
    proto.JWTBundlesResponse.serializeBinaryToWriter(this, writer);
    return writer.getResultBuffer();
};

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.JWTBundlesResponse} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.JWTBundlesResponse.serializeBinaryToWriter = function (message, writer) {
    var f = undefined;
    f = message.getBundlesMap(true);
    if (f && f.getLength() > 0) {
        f.serializeBinary(
            1,
            writer,
            jspb.BinaryWriter.prototype.writeString,
            jspb.BinaryWriter.prototype.writeBytes
        );
    }
};

/**
 * map<string, bytes> bundles = 1;
 * @param {boolean=} opt_noLazyCreate Do not create the map if
 * empty, instead returning `undefined`
 * @return {!jspb.Map<string,!(string|Uint8Array)>}
 */
proto.JWTBundlesResponse.prototype.getBundlesMap = function (opt_noLazyCreate) {
    return /** @type {!jspb.Map<string,!(string|Uint8Array)>} */ (
        jspb.Message.getMapField(this, 1, opt_noLazyCreate, null)
    );
};

/**
 * Clears values from the map. The map will be non-null.
 * @return {!proto.JWTBundlesResponse} returns this
 */
proto.JWTBundlesResponse.prototype.clearBundlesMap = function () {
    this.getBundlesMap().clear();
    return this;
};

if (jspb.Message.GENERATE_TO_OBJECT) {
    /**
     * Creates an object representation of this proto.
     * Field names that are reserved in JavaScript and will be renamed to pb_name.
     * Optional fields that are not set will be set to undefined.
     * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
     * For the list of reserved names please see:
     *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
     * @param {boolean=} opt_includeInstance Deprecated. whether to include the
     *     JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @return {!Object}
     */
    proto.ValidateJWTSVIDRequest.prototype.toObject = function (
        opt_includeInstance
    ) {
        return proto.ValidateJWTSVIDRequest.toObject(opt_includeInstance, this);
    };

    /**
     * Static version of the {@see toObject} method.
     * @param {boolean|undefined} includeInstance Deprecated. Whether to include
     *     the JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @param {!proto.ValidateJWTSVIDRequest} msg The msg instance to transform.
     * @return {!Object}
     * @suppress {unusedLocalVariables} f is only used for nested messages
     */
    proto.ValidateJWTSVIDRequest.toObject = function (includeInstance, msg) {
        var f,
            obj = {
                audience: jspb.Message.getFieldWithDefault(msg, 1, ''),
                svid: jspb.Message.getFieldWithDefault(msg, 2, ''),
            };

        if (includeInstance) {
            obj.$jspbMessageInstance = msg;
        }
        return obj;
    };
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.ValidateJWTSVIDRequest}
 */
proto.ValidateJWTSVIDRequest.deserializeBinary = function (bytes) {
    var reader = new jspb.BinaryReader(bytes);
    var msg = new proto.ValidateJWTSVIDRequest();
    return proto.ValidateJWTSVIDRequest.deserializeBinaryFromReader(
        msg,
        reader
    );
};

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.ValidateJWTSVIDRequest} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.ValidateJWTSVIDRequest}
 */
proto.ValidateJWTSVIDRequest.deserializeBinaryFromReader = function (
    msg,
    reader
) {
    while (reader.nextField()) {
        if (reader.isEndGroup()) {
            break;
        }
        var field = reader.getFieldNumber();
        switch (field) {
            case 1:
                var value = /** @type {string} */ (reader.readString());
                msg.setAudience(value);
                break;
            case 2:
                var value = /** @type {string} */ (reader.readString());
                msg.setSvid(value);
                break;
            default:
                reader.skipField();
                break;
        }
    }
    return msg;
};

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.ValidateJWTSVIDRequest.prototype.serializeBinary = function () {
    var writer = new jspb.BinaryWriter();
    proto.ValidateJWTSVIDRequest.serializeBinaryToWriter(this, writer);
    return writer.getResultBuffer();
};

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.ValidateJWTSVIDRequest} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.ValidateJWTSVIDRequest.serializeBinaryToWriter = function (
    message,
    writer
) {
    var f = undefined;
    f = message.getAudience();
    if (f.length > 0) {
        writer.writeString(1, f);
    }
    f = message.getSvid();
    if (f.length > 0) {
        writer.writeString(2, f);
    }
};

/**
 * optional string audience = 1;
 * @return {string}
 */
proto.ValidateJWTSVIDRequest.prototype.getAudience = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 1, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.ValidateJWTSVIDRequest} returns this
 */
proto.ValidateJWTSVIDRequest.prototype.setAudience = function (value) {
    return jspb.Message.setProto3StringField(this, 1, value);
};

/**
 * optional string svid = 2;
 * @return {string}
 */
proto.ValidateJWTSVIDRequest.prototype.getSvid = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 2, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.ValidateJWTSVIDRequest} returns this
 */
proto.ValidateJWTSVIDRequest.prototype.setSvid = function (value) {
    return jspb.Message.setProto3StringField(this, 2, value);
};

if (jspb.Message.GENERATE_TO_OBJECT) {
    /**
     * Creates an object representation of this proto.
     * Field names that are reserved in JavaScript and will be renamed to pb_name.
     * Optional fields that are not set will be set to undefined.
     * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
     * For the list of reserved names please see:
     *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
     * @param {boolean=} opt_includeInstance Deprecated. whether to include the
     *     JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @return {!Object}
     */
    proto.ValidateJWTSVIDResponse.prototype.toObject = function (
        opt_includeInstance
    ) {
        return proto.ValidateJWTSVIDResponse.toObject(
            opt_includeInstance,
            this
        );
    };

    /**
     * Static version of the {@see toObject} method.
     * @param {boolean|undefined} includeInstance Deprecated. Whether to include
     *     the JSPB instance for transitional soy proto support:
     *     http://goto/soy-param-migration
     * @param {!proto.ValidateJWTSVIDResponse} msg The msg instance to transform.
     * @return {!Object}
     * @suppress {unusedLocalVariables} f is only used for nested messages
     */
    proto.ValidateJWTSVIDResponse.toObject = function (includeInstance, msg) {
        var f,
            obj = {
                spiffeId: jspb.Message.getFieldWithDefault(msg, 1, ''),
                claims:
                    (f = msg.getClaims()) &&
                    google_protobuf_struct_pb.Struct.toObject(
                        includeInstance,
                        f
                    ),
            };

        if (includeInstance) {
            obj.$jspbMessageInstance = msg;
        }
        return obj;
    };
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.ValidateJWTSVIDResponse}
 */
proto.ValidateJWTSVIDResponse.deserializeBinary = function (bytes) {
    var reader = new jspb.BinaryReader(bytes);
    var msg = new proto.ValidateJWTSVIDResponse();
    return proto.ValidateJWTSVIDResponse.deserializeBinaryFromReader(
        msg,
        reader
    );
};

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.ValidateJWTSVIDResponse} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.ValidateJWTSVIDResponse}
 */
proto.ValidateJWTSVIDResponse.deserializeBinaryFromReader = function (
    msg,
    reader
) {
    while (reader.nextField()) {
        if (reader.isEndGroup()) {
            break;
        }
        var field = reader.getFieldNumber();
        switch (field) {
            case 1:
                var value = /** @type {string} */ (reader.readString());
                msg.setSpiffeId(value);
                break;
            case 2:
                var value = new google_protobuf_struct_pb.Struct();
                reader.readMessage(
                    value,
                    google_protobuf_struct_pb.Struct.deserializeBinaryFromReader
                );
                msg.setClaims(value);
                break;
            default:
                reader.skipField();
                break;
        }
    }
    return msg;
};

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.ValidateJWTSVIDResponse.prototype.serializeBinary = function () {
    var writer = new jspb.BinaryWriter();
    proto.ValidateJWTSVIDResponse.serializeBinaryToWriter(this, writer);
    return writer.getResultBuffer();
};

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.ValidateJWTSVIDResponse} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.ValidateJWTSVIDResponse.serializeBinaryToWriter = function (
    message,
    writer
) {
    var f = undefined;
    f = message.getSpiffeId();
    if (f.length > 0) {
        writer.writeString(1, f);
    }
    f = message.getClaims();
    if (f != null) {
        writer.writeMessage(
            2,
            f,
            google_protobuf_struct_pb.Struct.serializeBinaryToWriter
        );
    }
};

/**
 * optional string spiffe_id = 1;
 * @return {string}
 */
proto.ValidateJWTSVIDResponse.prototype.getSpiffeId = function () {
    return /** @type {string} */ (
        jspb.Message.getFieldWithDefault(this, 1, '')
    );
};

/**
 * @param {string} value
 * @return {!proto.ValidateJWTSVIDResponse} returns this
 */
proto.ValidateJWTSVIDResponse.prototype.setSpiffeId = function (value) {
    return jspb.Message.setProto3StringField(this, 1, value);
};

/**
 * optional google.protobuf.Struct claims = 2;
 * @return {?proto.google.protobuf.Struct}
 */
proto.ValidateJWTSVIDResponse.prototype.getClaims = function () {
    return /** @type{?proto.google.protobuf.Struct} */ (
        jspb.Message.getWrapperField(this, google_protobuf_struct_pb.Struct, 2)
    );
};

/**
 * @param {?proto.google.protobuf.Struct|undefined} value
 * @return {!proto.ValidateJWTSVIDResponse} returns this
 */
proto.ValidateJWTSVIDResponse.prototype.setClaims = function (value) {
    return jspb.Message.setWrapperField(this, 2, value);
};

/**
 * Clears the message field making it undefined.
 * @return {!proto.ValidateJWTSVIDResponse} returns this
 */
proto.ValidateJWTSVIDResponse.prototype.clearClaims = function () {
    return this.setClaims(undefined);
};

/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.ValidateJWTSVIDResponse.prototype.hasClaims = function () {
    return jspb.Message.getField(this, 2) != null;
};

goog.object.extend(exports, proto);
