// @universe-design/react/src/date-picker/cell/index.tsx
import React from 'react';
import clsx from 'clsx';
import { useCssScope } from '../../config-provider/hooks';
import type { ICellStatus } from './Range.class';
import { Range } from './Range.class';

import './index.less';

/**
 * DateCell 包含三个部分
 * - DateCellRangeBG，range 背景
 * - DateCellRangePreviewBG，range 预选背景
 * - DateCellInner，真正的日期元胞
 *
 * range 背景会有 start 和 end 的概念，会出现半圆形
 * range preview 和 range 是有叠加需求的
 */
export interface IDateCellProps extends React.ComponentPropsWithoutRef<'div'>, ICellStatus {
  /**
   * 之前一个 cell 的节点状态
   */
  prevCellStatus?: ICellStatus;
  /**
   * 之后一个 cell 的节点状态
   */
  nextCellStatus?: ICellStatus;
  colIndex?: number;
  colNum?: number;
  mode?: 'text' | 'cell';
}

const usePickerCellCssScope = () => useCssScope('picker__cell');

const DateCell = React.forwardRef<HTMLDivElement, IDateCellProps>(
  (
    {
      children,
      isSelected = false,
      isDisabled = false,
      isToday = false,
      isOutOfView = false,
      isRangePreview = false,
      isRange = false,
      className,
      prevCellStatus,
      nextCellStatus,
      mode = 'cell',
      colIndex,
      colNum,
      ...restProps
    },
    ref,
  ) => {
    const cssScope = usePickerCellCssScope();
    const rangeCssScope = useCssScope('picker__cell__range');

    if (mode === 'text') {
      return (
        <div ref={ref} className={clsx(cssScope(), className)} {...restProps}>
          <div className={clsx(cssScope('__inner'), cssScope('__inner-text'))}>{children}</div>
        </div>
      );
    }

    if (!prevCellStatus || !nextCellStatus || colIndex === undefined || colNum === undefined) {
      return null;
    }

    /**
     * 这段逻辑的解释在如下文档中
     * https://bytedance.feishu.cn/docx/doxcn9kqBZVTrbolR9ilVUzsR1X
     */

    // 全局截断
    const isRowStart = colIndex === 0;
    const isRowEnd = colIndex === colNum - 1;

    const cellStatus: ICellStatus = {
      isSelected,
      isDisabled,
      isToday,
      isOutOfView,
      isRangePreview,
      isRange,
    };

    const {
      cellColor: rangeCellColor,
      beforeColor: rangeBeforeColor,
      afterColor: rangeAfterColor,
    } = new Range(prevCellStatus, cellStatus, nextCellStatus);

    const rangeBeforeBGCls = (() => {
      if (isRowStart) {
        return;
      }
      return rangeBeforeColor ? rangeCssScope(`-bg-${rangeBeforeColor}`) : undefined;
    })();

    const rangeAfterBGCls = (() => {
      if (isRowEnd) {
        return;
      }
      return rangeAfterColor ? rangeCssScope(`-bg-${rangeAfterColor}`) : undefined;
    })();

    // Range 跟前面一段相接的部分
    const beforeRangeNode = <div className={clsx(rangeCssScope('__before'), rangeBeforeBGCls)} />;

    // Range 跟后面一段相接的部分
    const afterRangeNode = <div className={clsx(rangeCssScope('__after'), rangeAfterBGCls)} />;

    return (
      <>
        <div
          ref={ref}
          className={clsx(
            cssScope(),
            isSelected && cssScope('-selected'),
            isOutOfView && cssScope('-outOfView'),
            isDisabled && cssScope('-disabled'),
            isToday && cssScope('-today'),
            isRange && cssScope('-range'),
            isRangePreview && cssScope('-rangePreview'),
            className,
          )}
          {...restProps}
        >
          {beforeRangeNode}
          {afterRangeNode}
          {/* Range 中间的部分，可能跟前面或者后面一段相接，呈现一个胶囊形状 */}
          <div
            className={clsx(cssScope('__range'), rangeCellColor ? rangeCssScope(`-bg-${rangeCellColor}`) : undefined)}
          >
            <div className={cssScope('__inner')}>{children}</div>
          </div>
        </div>
        {!isRowEnd && <div className={clsx(cssScope('__space'), rangeAfterBGCls)} />}
      </>
    );
  },
);

export const useDateCellInteractiveAreaCls = () => usePickerCellCssScope()('-interactive-area');

export default DateCell;
