// @universe-design/react/src/date-picker/cell/Range.class.ts
export interface ICellStatus {
  /**
   * 是否是今天
   */
  isToday?: boolean;
  /**
   * 是否是禁用状态
   */
  isDisabled?: boolean;
  /**
   * 是否是 range value 预选值
   */
  isRangePreview?: boolean;
  /**
   * 是否是 range value 选中值
   */
  isRange?: boolean;
  /**
   * 是否选中值
   */
  isSelected?: boolean;
  /**
   * 是否是 panel 以外的值
   */
  isOutOfView?: boolean;
}

enum RangeType {
  OutOfView,
  InView,
  Selected,
  OutOfSelected,
  Preview,
  Disabled,
}

export type RangeColor = 'selected' | 'preview' | 'selected-preview' | 'disabled';

export class Range {
  private _prev: ICellStatus;
  private _cur: ICellStatus;
  private _next: ICellStatus;
  private _startEnd: Map<RangeType, { start: boolean; end: boolean; in: boolean }>;

  constructor(_prev: ICellStatus, _cur: ICellStatus, _next: ICellStatus) {
    this._prev = _prev;
    this._cur = _cur;
    this._next = _next;
    this._startEnd = new Map([
      [
        RangeType.OutOfView,
        {
          start: Boolean(this._cur.isOutOfView && !this._prev.isOutOfView),
          end: Boolean(this._cur.isOutOfView && !this._next.isOutOfView),
          in: Boolean(this._cur.isOutOfView),
        },
      ],
      [
        RangeType.InView,
        {
          start: Boolean(!this._cur.isOutOfView && this._prev.isOutOfView),
          end: Boolean(!this._cur.isOutOfView && this._next.isOutOfView),
          in: !this._cur.isOutOfView,
        },
      ],
      [
        RangeType.Selected,
        {
          start: Boolean(this._cur.isRange && !this._prev.isRange),
          end: Boolean(this._cur.isRange && !this._next.isRange),
          in: Boolean(this._cur.isRange),
        },
      ],
      [
        RangeType.OutOfSelected,
        {
          start: Boolean(!this._cur.isRange && this._prev.isRange),
          end: Boolean(!this._cur.isRange && this._next.isRange),
          in: !this._cur.isRange,
        },
      ],
      [
        RangeType.Preview,
        {
          start: Boolean(this._cur.isRangePreview && !this._prev.isRangePreview),
          end: Boolean(this._cur.isRangePreview && !this._next.isRangePreview),
          in: Boolean(this._cur.isRangePreview),
        },
      ],
      [
        RangeType.Disabled,
        {
          start: Boolean(this._cur.isDisabled && !this._prev.isDisabled),
          end: Boolean(this._cur.isDisabled && !this._next.isDisabled),
          in: Boolean(this._cur.isDisabled),
        },
      ],
    ]);
  }

  get cellColor(): RangeColor | undefined {
    const typeSet = new Set<RangeType>();
    for (const [rangeType, se] of this._startEnd.entries()) {
      if (se.in) {
        typeSet.add(rangeType);
      }
    }
    if (this._startEnd.get(RangeType.OutOfView)?.in) {
      return;
    }
    return Range.rangeTypes2rangeColor(typeSet);
  }

  get beforeColor(): RangeColor | undefined {
    const typeSet = new Set<RangeType>();
    for (const [rangeType, se] of this._startEnd.entries()) {
      if (se.in && !se.start) {
        typeSet.add(rangeType);
      }
    }
    if (this._startEnd.get(RangeType.OutOfView)?.in || this._startEnd.get(RangeType.InView)?.start) {
      return;
    }
    if (this._startEnd.get(RangeType.Selected)?.start || this._startEnd.get(RangeType.OutOfSelected)?.start) {
      typeSet.delete(RangeType.Disabled);
    }
    return Range.rangeTypes2rangeColor(typeSet);
  }

  get afterColor(): RangeColor | undefined {
    const typeSet = new Set<RangeType>();
    for (const [rangeType, se] of this._startEnd.entries()) {
      if (se.in && !se.end) {
        typeSet.add(rangeType);
      }
    }
    // 特化逻辑，view range 开头结尾前后需要截断所有
    if (this._startEnd.get(RangeType.OutOfView)?.in || this._startEnd.get(RangeType.InView)?.end) {
      return;
    }
    // 特化逻辑，selected range 开头结尾的前后需要截断 disabled
    if (this._startEnd.get(RangeType.Selected)?.end || this._startEnd.get(RangeType.OutOfSelected)?.end) {
      typeSet.delete(RangeType.Disabled);
    }
    return Range.rangeTypes2rangeColor(typeSet);
  }

  static rangeTypes2rangeColor(typeSet: Set<RangeType>): RangeColor | undefined {
    if (typeSet.has(RangeType.Disabled)) {
      return 'disabled';
    }
    if (typeSet.has(RangeType.Preview) && typeSet.has(RangeType.Selected)) {
      return 'selected-preview';
    }
    if (typeSet.has(RangeType.Preview)) {
      return 'preview';
    }
    if (typeSet.has(RangeType.Selected)) {
      return 'selected';
    }
  }
}
