// @universe-design/react/es/date-picker/panels/month-body/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React from 'react';
import PanelBody from '../base/Body';
import { BodyConfig } from '../base/config';
import { useCssScope, useLocaleReceiver } from '../../../config-provider/hooks';
import { useDateCellInteractiveAreaCls } from '../../cell';
import { useDateCalendarUIContext } from '../DateCalendar/DateCalendarUIContext';
import { isSelectCell } from '../DateCalendar/utils';
function MonthPanel() {
  const locale = useLocaleReceiver('datePicker');
  const {
    shouldCellDisabled,
    cellRenderer,
    pickerValue,
    selectedDates,
    selectedRange,
    previewDate
  } = useDateCalendarUIContext();
  const prefixCls = useCssScope('picker')();
  const dateCellInteractiveAreaCls = useDateCellInteractiveAreaCls();
  return /*#__PURE__*/React.createElement(PanelBody, _extends({}, BodyConfig.month, {
    cellClassName: `${`${prefixCls}-month-panel`}-cell`,
    baseDate: pickerValue.clone().startOf('year'),
    getCellStatus: date => {
      const isRangePreview = previewDate ? previewDate(date) : false;
      const [startDate = null, endDate = null] = selectedRange || [];
      return {
        isDisabled: shouldCellDisabled(date, 'month'),
        isRange: date.isSameOrAfter(startDate, 'month') && date.isSameOrBefore(endDate, 'month'),
        isSelected: isSelectCell({
          selectedDates,
          selectedRange,
          mode: 'month',
          date
        }),
        isRangePreview
      };
    },
    getCellDate: (d, offset) => d.clone().add(offset, 'month'),
    dateRender: d => {
      const node = /*#__PURE__*/React.createElement("div", {
        className: dateCellInteractiveAreaCls
      }, locale.monthsMin[d.month()]);
      if (typeof cellRenderer === 'function') {
        return cellRenderer(d, {
          viewMode: 'month',
          defaultNode: node
        });
      }
      return node;
    }
  }));
}
export default MonthPanel;
