// @byted-service/tcc-general/lib/bconfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const node_fetch_1 = tslib_1.__importDefault(require("node-fetch"));
const error_1 = tslib_1.__importStar(require("./error"));
const metrics_1 = tslib_1.__importDefault(require("@byted-service/metrics"));
const consul_1 = require("@byted-service/consul");
const consts_1 = require("./consts");
const metricsClient = new metrics_1.default({ defaultPrefix: 'tcc.bconfig.proxy' });
class BConfig {
    constructor(logger, cluster = consts_1.TCC_B_CONFIG_PROXY_PSM, getTimeout = consts_1.DEFAULT_M_GET_TIMEOUT, bconfigPSM = consts_1.DEFAULT_CLUSTER_NAME) {
        this.consul = new consul_1.Consul();
        this.cluster = cluster;
        this.bconfigPSM = bconfigPSM;
        this.getTimeout = getTimeout;
        this.logger = logger;
    }
    /**
     * 从服务端获取单个数据
     * @param key
     * @param versionCode
     */
    async getRemote(key, versionCode) {
        return await this.getFromServer(key, versionCode);
    }
    async multiGetRemote(nodes) {
        const t0 = Date.now();
        try {
            const addr = await this.lookupBConfigProxy();
            const res = await (0, node_fetch_1.default)(`http://${addr.uri}/v2/multi_nodes`, Object.assign(Object.assign({}, consts_1.DEFAULT_FETCH_OPTIONS), { method: 'POST', body: JSON.stringify({ nodes }), timeout: this.getTimeout }));
            this.emit(t0, null, 'multi_nodes');
            return await this.decodeNodesResp(res);
        }
        catch (error) {
            this.emit(t0, error, 'multi_nodes');
            throw new error_1.default(error_1.TccErrorCode.errRemoteFailed, error.message);
        }
    }
    async getFromServer(key, versionCode) {
        const t0 = Date.now();
        let res;
        try {
            const addr = await this.lookupBConfigProxy();
            const url = `http://${addr.uri}/v2/nodes/${key}?v=${versionCode}`;
            // 误报
            res = await (0, node_fetch_1.default)(url, Object.assign(Object.assign({}, consts_1.DEFAULT_FETCH_OPTIONS), { method: 'GET', timeout: this.getTimeout }));
            this.emit(t0, null, 'node');
        }
        catch (error) {
            this.emit(t0, error, 'node');
            this.logger.warn('[tcc-general] get from remote failed, err: %o', error);
            throw new error_1.default(error_1.TccErrorCode.errRemoteFailed, error.message);
        }
        return this.decodeNodeResp(res);
    }
    async decodeNodeResp(res) {
        var _a;
        const status = res.status;
        // not valid code(200/304/400)
        if (
        // tslint:disable-next-line: triple-equals
        status != consts_1.HttpStatusCode.statusNotModified &&
            // tslint:disable-next-line: triple-equals
            status != consts_1.HttpStatusCode.statusBadRequest &&
            // tslint:disable-next-line: triple-equals
            status != consts_1.HttpStatusCode.statusOK) {
            this.logger.warn('[tcc] remote get has response status code %d', status);
            throw new error_1.default(error_1.TccErrorCode.errRemoteFailed);
        }
        if (status === consts_1.HttpStatusCode.statusNotModified) {
            throw new error_1.default(error_1.TccErrorCode.errNotModified);
        }
        let body;
        try {
            body = await res.json();
        }
        catch (error) {
            throw new error_1.default(error_1.TccErrorCode.errOther, error.message);
        }
        if (status === consts_1.HttpStatusCode.statusBadRequest) {
            // tslint:disable-next-line: triple-equals
            if (body.code == consts_1.HttpStatusCode.respErrCodeNodeNotFound) {
                throw new error_1.default(error_1.TccErrorCode.errKeyNotFound);
            }
            let level = 'error';
            /* istanbul ignore next */
            if ((_a = body === null || body === void 0 ? void 0 : body.message) === null || _a === void 0 ? void 0 : _a.includes('context deadline exceeded')) {
                level = 'warn';
            }
            this.logger[level]('[tcc-general] get has response err_code: %s, err_msg: %s', body.code, body.message);
            throw new error_1.default(error_1.TccErrorCode.errRemoteFailed);
        }
        return body.data;
    }
    async decodeNodesResp(res) {
        const status = res.status;
        // tslint:disable-next-line: triple-equals
        if (status != consts_1.HttpStatusCode.statusOK) {
            this.logger.warn('[tcc-general]remote multi-get has response not ok, status code %s', status);
            throw new error_1.default(error_1.TccErrorCode.errRemoteFailed);
        }
        let body;
        try {
            body = await res.json();
        }
        catch (error) {
            throw new error_1.default(error_1.TccErrorCode.errOther, error.message);
        }
        const retDatas = [];
        body.data.forEach((nodeData) => {
            var _a;
            let err = null;
            let data = null;
            switch (nodeData.code) {
                case consts_1.MultiResCode.respErrCodeOK:
                    data = nodeData.data;
                    break;
                case consts_1.MultiResCode.respErrCodeNotModified:
                    err = new error_1.default(error_1.TccErrorCode.errNotModified);
                    break;
                case consts_1.MultiResCode.respErrCodeNodeNotFound:
                    err = new error_1.default(error_1.TccErrorCode.errKeyNotFound);
                    break;
                default:
                    /* istanbul ignore next */
                    if (!((_a = nodeData === null || nodeData === void 0 ? void 0 : nodeData.message) === null || _a === void 0 ? void 0 : _a.includes('context deadline exceeded'))) {
                        this.logger.warn('[tcc-general]remote multi-get failed, node_name: %s, err_code: %s, err_msg: %s', nodeData.node, nodeData.code, nodeData.message);
                    }
                    err = new error_1.default(error_1.TccErrorCode.errRemoteFailed);
            }
            retDatas.push({
                nodeName: nodeData.node,
                err: err || undefined,
                nodeData: data || undefined,
            });
        });
        return retDatas;
    }
    async emit(t0, err, api) {
        const cost = Date.now() - t0;
        const status = err ? 'failed' : 'success';
        const tags = {
            api,
            client: 'js',
        };
        await metricsClient.emitCounter(`get.${status}.throughput`, 1, '', tags);
        await metricsClient.emitTimer(`get.${status}.throughput`, cost, '', tags);
    }
    async lookupBConfigProxy() {
        let endpoint;
        let err;
        try {
            endpoint = await (0, consul_1.random)(consts_1.TCC_B_CONFIG_PROXY_PSM, {}, this.consul);
        }
        catch (error) {
            err = error;
        }
        // falblack
        if (!endpoint) {
            endpoint = await (0, consul_1.random)(consts_1.TCC_BYTEDANCE_CONFIG_PROXY_PSM, {}, this.consul);
        }
        /* istanbul ignore if */
        if (!endpoint) {
            throw err;
        }
        return endpoint;
    }
}
exports.default = BConfig;
