// @byted-service/tcc-general/lib/emiter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const consts_1 = require("./consts");
const error_1 = require("./error");
const metrics_1 = tslib_1.__importDefault(require("@byted-service/metrics"));
// tslint:disable-next-line:no-var-requires
const pkg = require('../package.json');
var EmitStatus;
(function (EmitStatus) {
    EmitStatus["statusSucc"] = "success";
    EmitStatus["statusErr"] = "failed";
    EmitStatus["statusNotFound"] = "notFound";
})(EmitStatus || (EmitStatus = {}));
const DEFAULT_NOT_FOUND_KEY = 'notFoundKey';
const EMIT_CACHE_INTERVAL = 10000;
const metricsClient = new metrics_1.default({ defaultPrefix: 'tcc' });
class MetricsEmiter {
    constructor(serviceName, cluster) {
        this.cache = new Map();
        this.serviceName = serviceName;
        this.cluster = cluster;
        this.notFoundItem = { succ: 0, err: 0, notFound: 0 };
        this.cache.set(DEFAULT_NOT_FOUND_KEY, this.notFoundItem);
        this.emitCache();
    }
    async emit(err, configKey, getType) {
        let status = EmitStatus.statusSucc;
        if ((err === null || err === void 0 ? void 0 : err.code) === error_1.TccErrorCode.errKeyNotFound) {
            status = EmitStatus.statusNotFound;
        }
        else if (err) {
            status = EmitStatus.statusErr;
        }
        if (getType !== consts_1.GetType.getTypeCache) {
            await this.doEmit({
                status,
                nodeName: configKey,
                getType,
                count: 1,
            });
            return;
        }
        let item = this.cache.get(configKey);
        if (!item) {
            if (status === EmitStatus.statusNotFound) {
                item = this.notFoundItem;
            }
            else {
                item = { succ: 0, err: 0, notFound: 0 };
                this.cache.set(configKey, item);
            }
        }
        switch (status) {
            case EmitStatus.statusSucc:
                item.succ++;
                break;
            case EmitStatus.statusErr:
                item.err++;
                break;
            case EmitStatus.statusNotFound:
                item.notFound++;
                break;
        }
    }
    async emitCache() {
        while (true) {
            await (0, consts_1.sleep)(EMIT_CACHE_INTERVAL);
            try {
                for (const [k, v] of this.cache) {
                    const emitBaseData = {
                        nodeName: k,
                        getType: consts_1.GetType.getTypeCache,
                    };
                    const succCnt = v.succ;
                    const errCnt = v.err;
                    const notFoundCnt = v.notFound;
                    await this.doEmit(Object.assign(Object.assign({}, emitBaseData), { status: EmitStatus.statusSucc, count: succCnt }));
                    await this.doEmit(Object.assign(Object.assign({}, emitBaseData), { status: EmitStatus.statusErr, count: errCnt }));
                    await this.doEmit(Object.assign(Object.assign({}, emitBaseData), { status: EmitStatus.statusNotFound, count: notFoundCnt }));
                    v.succ = v.succ - succCnt;
                    v.err = v.err - errCnt;
                    v.notFound = v.notFound - notFoundCnt;
                }
            }
            catch (error) {
                // do nothing
            }
        }
    }
    async doEmit({ status, nodeName, getType, count, }) {
        if (count <= 0) {
            /* istanbul ignore next */
            return;
        }
        let code = '0';
        if (status === EmitStatus.statusErr) {
            code = '1';
        }
        if (status === EmitStatus.statusNotFound) {
            status = EmitStatus.statusSucc;
            code = '100';
        }
        const key = `general_client.${this.serviceName}.get_config.${status}`;
        const tags = {
            bconfig_cluster: this.cluster,
            config_node: nodeName,
            version: pkg.version,
            language: 'js',
            code,
            get_type: getType,
        };
        await metricsClient.emitCounter(key, count, '', tags); // nolint
    }
}
exports.default = MetricsEmiter;
