// @universe-design/styles/cjs/core/index.dark.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.coretoken_video_component_drag_bar_bg_normal = exports.coretoken_settings_shadow_keyboard_down_normal = exports.coretoken_settings_notice_banner_bg_general = exports.coretoken_settings_color_tag_text = exports.coretoken_settings_color_tag_bg_up = exports.coretoken_settings_color_tag_bg_down = exports.coretoken_settings_bg_keyboard_down_normal = exports.coretoken_newtab_color_app_bg = exports.coretoken_navigation_icon_bg_hover = exports.coretoken_navigation_btn_bg_normal = exports.coretoken_information_bg_bell = exports.coretoken_groupshare_bg_placeholder = exports.coretoken_card_float = exports.coretoken_all_watermark_general = void 0;
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
exports.coretoken_all_watermark_general = 'rgba(117, 117, 117, 0.08)';
exports.coretoken_card_float = '#373737';
exports.coretoken_groupshare_bg_placeholder = 'linear-gradient(90deg, #4c88ff 0%, #2e65d1 100%)';
exports.coretoken_information_bg_bell = '#373737';
exports.coretoken_navigation_btn_bg_normal = 'rgba(240, 240, 240, 0.05)';
exports.coretoken_navigation_icon_bg_hover = '#505050';
exports.coretoken_newtab_color_app_bg = 'rgba(255, 255, 255, 0.1)';
exports.coretoken_settings_bg_keyboard_down_normal = '#373737';
exports.coretoken_settings_color_tag_bg_down = '#fcd456';
exports.coretoken_settings_color_tag_bg_up = '#ffeaa3';
exports.coretoken_settings_color_tag_text = '#63470f';
exports.coretoken_settings_notice_banner_bg_general = '#373737';
exports.coretoken_settings_shadow_keyboard_down_normal = '#0a0a0a';
exports.coretoken_video_component_drag_bar_bg_normal = '#0a0a0a';
