// @universe-design/react/src/time-picker/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Basic from '../demo/0.basic';
import C1Use12Hours from '../demo/1.use12Hours';
import C2Second from '../demo/2.second';
import C3Disable from '../demo/3.disable';
import C4Step from '../demo/4.step';
import C5Custom from '../demo/5.custom';
import C6Icon from '../demo/6.icon';
if (!(C0Basic as Story).storyName) {
  (C0Basic as Story).storyName = '0. 类型';
  (C0Basic as Story).parameters = {
    docs: {
      description: {
        story:
          '- 当下拉选项的时间间隔为大于等于 15min 时建议使用时和分合并选项的时间选择器 (mode="timeCombine")。\n- 当下拉选项的时间间隔为小于 15min 时建议使用时和分独立选项的时间选择器。\n\n',
      },
    },
  };
}
if (!(C1Use12Hours as Story).storyName) {
  (C1Use12Hours as Story).storyName = '0. 12小时制';
  (C1Use12Hours as Story).parameters = {
    docs: {
      description: {
        story: '12 小时制的时间选择器，format 为 `hh:mm`。\n\n',
      },
    },
  };
}
if (!(C2Second as Story).storyName) {
  (C2Second as Story).storyName = '0. 选择秒';
  (C2Second as Story).parameters = {
    docs: {
      description: {
        story: '时间选择器支持选择精确到秒，format 为 `HH:mm:ss`。注意，mode="timeCombine"时不可选择秒。\n\n',
      },
    },
  };
}
if (!(C3Disable as Story).storyName) {
  (C3Disable as Story).storyName = '0. 禁用时间';
  (C3Disable as Story).parameters = {
    docs: {
      description: {
        story: '禁用时间。\n\n',
      },
    },
  };
}
if (!(C4Step as Story).storyName) {
  (C4Step as Story).storyName = '0. 自定义时间间隔';
  (C4Step as Story).parameters = {
    docs: {
      description: {
        story:
          '- 可以使用 `hourStep` `minuteStep` `secondStep` 按步长展示可选的时分秒。\n- 在 `mode="timeCombine"` 时， 可以使用 timeStep 定义时间间隔。\n\n',
      },
    },
  };
}
if (!(C5Custom as Story).storyName) {
  (C5Custom as Story).storyName = '0. 自定义选择项';
  (C5Custom as Story).parameters = {
    docs: {
      description: {
        story: '在 `mode="timeCombine"` 时，可以自定义选择项。\n\n',
      },
    },
  };
}
if (!(C6Icon as Story).storyName) {
  (C6Icon as Story).storyName = '0. 自定义图标';
  (C6Icon as Story).parameters = {
    docs: {
      description: {
        story: '自定义选择框后缀图标。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/TimePicker/官网示例',
  tags: ['autodocs'],
};

export { C0Basic, C1Use12Hours, C2Second, C3Disable, C4Step, C5Custom, C6Icon };
