// @universe-design/react/src/slider/handle.tsx
import clsx from 'clsx';
import type { FC } from 'react';
import React, { useEffect, useRef, useState } from 'react';
import { useCssScope } from '../config-provider/hooks';
import type { TooltipPlacement } from '../tooltip';
import { Tooltip } from '../tooltip';
import { isFunction } from '../util';
import type { UdTriggerRef } from '../shared/components';
import { useEventCallback } from '../hooks/useEventCallback';

export interface SliderHandleProps {
  max: number;
  min: number;
  value?: number;
  vertical?: boolean;
  dragging?: boolean;
  tipFormatter?: null | ((value?: number) => React.ReactNode);
  getTooltipPopupContainer?: (triggerNode: HTMLElement) => HTMLElement;
  tooltipVisible?: boolean;
  tooltipPlacement?: TooltipPlacement;
}

export const SliderHandle: FC<SliderHandleProps> = props => {
  const {
    max,
    min,
    value,
    vertical,
    dragging,
    tooltipVisible: tooltipVisibleInProps,
    tooltipPlacement,
    tipFormatter,
    getTooltipPopupContainer,
  } = props;

  const cssScope = useCssScope('slider');
  const unSetVisible = typeof tooltipVisibleInProps === 'undefined';
  const renderTooltip = unSetVisible || tooltipVisibleInProps;
  const [toolTipVisible, setToolTipVisible] = useState<boolean>();
  const udTooltipRef = useRef<UdTriggerRef | null>(null);

  useEffect(() => {
    if (unSetVisible && typeof dragging !== 'undefined') {
      setToolTipVisible(dragging);
    } else {
      setToolTipVisible(tooltipVisibleInProps);
    }
  }, [unSetVisible, tooltipVisibleInProps, dragging]);

  useEffect(() => {
    // 1. 由于 `forcePopupAlign` 本身就是做了 throttle 的，所以我们用 `useEffect` 就可以了
    // 2. deps 数组先用 [value]，这种情况下如果 `min` 和 `max` 动态变化并不会触发 `forcePopupAlign`，但是大概率用户不会有拖动导致 `value` 变化的时候动态变化 `min` 和 `max` 的用法，先这样改，看看用户反馈
    udTooltipRef.current?.forcePopupAlign();
  }, [value]);

  const handleTooltipVisibleChange = (visible: boolean) => {
    if (unSetVisible) {
      if (typeof dragging === 'undefined') {
        setToolTipVisible(visible);
      } else {
        setToolTipVisible(visible || dragging);
      }
    }
  };

  const handleHolderMouseDown = useEventCallback<React.MouseEventHandler<HTMLDivElement>>(ev => {
    // 解决 Safari 上拖动时光标从 pointer 变成 text 的问题
    ev.preventDefault();
  });

  const renderHandle = () => {
    const v = value ? Math.round(((value - min) / (max - min)) * 100) : 0;
    const style = vertical ? { bottom: `${v}%` } : { left: `${v}%` };
    const handleCls = clsx(cssScope('__handle'), {
      [cssScope('__handle--active')]: dragging,
    });
    const title = isFunction(tipFormatter) ? tipFormatter(value) : `${value}`;
    if (renderTooltip && tipFormatter !== null) {
      return (
        <Tooltip
          title={title}
          visible={toolTipVisible}
          onVisibleChange={handleTooltipVisibleChange}
          getPopupContainer={getTooltipPopupContainer}
          placement={tooltipPlacement}
          udTriggerRef={udTooltipRef}
        >
          <div style={style} className={cssScope('__handle-holder')} onMouseDown={handleHolderMouseDown}>
            <div className={handleCls} />
          </div>
        </Tooltip>
      );
    }
    return (
      <div style={style} className={cssScope('__handle-holder')} onMouseDown={handleHolderMouseDown}>
        <div className={handleCls} />
      </div>
    );
  };
  return <>{renderHandle()}</>;
};
