// @universe-design/react/src/slider/marks.tsx
import clsx from 'clsx';
import type { FC } from 'react';
import React, { isValidElement } from 'react';
import { useCssScope } from '../config-provider/hooks';
import type { SliderMarksOption } from './interface';
import { Text } from '../text';

export interface SliderMarksProps {
  max: number;
  min: number;
  marks?: SliderMarksOption;
  reverse?: boolean;
  vertical?: boolean;
  onMarkClick: (value: number) => void;
}

export const SliderMarks: FC<SliderMarksProps> = props => {
  const { max, min, marks, reverse, vertical, onMarkClick } = props;

  const cssScope = useCssScope('slider');

  const markCls = clsx(cssScope('__marks'), {
    [cssScope('__marks--vertical')]: vertical,
  });

  const handleMarkClick = (e: React.MouseEvent<HTMLDivElement, MouseEvent>, value: number) => {
    e.stopPropagation();
    onMarkClick(value);
  };

  const renderMarks = () => {
    if (marks) {
      const keys = Object.keys(marks);
      return keys
        .map(parseFloat)
        .sort((a, b) => a - b)
        .map(key => {
          const item = marks[key];
          const isObject = typeof item === 'object' && !isValidElement(item);
          const label = isObject ? (item as any).label : item;
          const style = isObject ? (item as any).style : undefined;
          if (!label && label !== 0) {
            return null;
          }
          const cls = clsx(cssScope('__marks-item'));
          const point = key;
          const mergedStyle = {
            ...style,
            ...(vertical
              ? {
                  [reverse ? 'top' : 'bottom']: `${((point - min) / (max - min)) * 100}%`,
                }
              : {
                  [reverse ? 'right' : 'left']: `${((point - min) / (max - min)) * 100}%`,
                }),
          };

          return (
            <Text
              as="div"
              variant="body2"
              key={key}
              style={mergedStyle}
              className={cls}
              onMouseDown={e => handleMarkClick(e, point)}
            >
              {label}
            </Text>
          );
        });
    }
    return null;
  };

  return marks ? <div className={markCls}>{renderMarks()}</div> : null;
};
