// @universe-design/react/src/slider/steps.tsx
import clsx from 'clsx';
import type { FC } from 'react';
import React from 'react';
import { useCssScope } from '../config-provider/hooks';
export interface SliderMarksProps {
  max: number;
  min: number;
  range?: boolean;
  value?: number;
  rangeValue?: [number, number];
  dots?: number[];
  reverse?: boolean;
  vertical?: boolean;
  included?: boolean;
}

export const SliderSteps: FC<SliderMarksProps> = props => {
  const { max, min, range, dots = [], value, rangeValue, reverse, vertical, included } = props;

  const cssScope = useCssScope('slider');
  const stepsCls = clsx(cssScope('__steps'));

  const renderMarks = () => {
    if (dots) {
      return dots
        .sort((a, b) => a - b)
        .map(item => {
          const isActive = range
            ? rangeValue && item >= rangeValue[0] && item <= rangeValue[1]
            : value && item <= value;
          const cls = clsx(cssScope('__steps-item'), {
            [cssScope('__steps-item--active')]: isActive && included,
          });
          const point = item;
          const style = vertical
            ? {
                [reverse ? 'top' : 'bottom']: `${((point - min) / (max - min)) * 100}%`,
              }
            : {
                [reverse ? 'right' : 'left']: `${((point - min) / (max - min)) * 100}%`,
              };
          return <div key={`ud-slider-step-${item}`} style={style} className={cls} />;
        });
    }
    return null;
  };

  return dots && dots.length > 0 ? <div className={stepsCls}>{renderMarks()}</div> : null;
};
