// @byted-service/netpool/lib/udp/operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UDPOperation = void 0;
const dgram_1 = require("dgram");
const util_1 = require("util");
const helper_1 = require("../helper");
const consts_1 = require("../consts");
const idCreator = (0, helper_1.genIdCreator)('udpsocket');
/**
 * Global clock
 * 全局时钟
 */
let now = Date.now();
setInterval(() => (now = Date.now()), 1000).unref();
class UDPOperation {
    constructor(options) {
        this.id = '-';
        this.connectOptions = options.connectOptions;
        this.connectTimeout =
            typeof options.connectTimeout === 'number' ||
                options.connectTimeout === null
                ? options.connectTimeout
                : consts_1.defaultUDPPoolOptions.connectTimeout;
        this.bindOptions = options.bindOptions;
        this.maxSocketSurvivalTime =
            options.maxSocketSurvivalTime ||
                consts_1.defaultUDPPoolOptions.maxSocketSurvivalTime;
        this.maxSocketAge =
            options.maxSocketAge || consts_1.defaultUDPPoolOptions.maxSocketAge;
        this.customCreate = options.create;
        this.customDestroy = options.destroy;
        this.customValidate = options.validate;
        this.customReset = options.reset;
    }
    create() {
        return new Promise((resolve, reject) => {
            const socket = (0, dgram_1.createSocket)(this.connectOptions);
            socket.id = idCreator();
            socket.age = 0;
            socket.createdAt = now;
            socket.closed = false;
            socket.bind(Object.assign({ 
                /**
                 * https://nodejs.org/dist/latest-v12.x/docs/api/dgram.html#dgram_socket_bind_options_callback
                 *
                 * Exclusive must be `true` on in cluster mode: true, otherwise the following errors may be caused ( it is not clear if it is a Node.js bug or feature )
                 * 在 cluster 模式下必须开启 exclusive: true，否则可能会造成以下错误 ( 暂时不清楚是否是 Node.js 的 bug )
                 *
                 * ```sh
                 * Error [ERR_INTERNAL_ASSERTION]: This is caused by either a bug in Node.js or incorrect usage of Node.js internals.
                 * Please open an issue with this stack trace at https://github.com/nodejs/node/issues
                 *
                 *    at assert (internal/assert.js:14:11)
                 *    at SharedHandle.add (internal/cluster/shared_handle.js:28:3)
                 *    at queryServer (internal/cluster/master.js:309:10)
                 *    at Worker.onmessage (internal/cluster/master.js:249:5)
                 *    at ChildProcess.onInternalMessage (internal/cluster/utils.js:47:8)
                 *    at ChildProcess.emit (events.js:326:22)
                 *    at emit (internal/child_process.js:906:12)
                 *    at processTicksAndRejections (internal/process/task_queues.js:81:21) {
                 *    code: 'ERR_INTERNAL_ASSERTION'
                 * }
                 * ```
                 *
                 * The specific reason is caused by cluster mode + shared UDP handle
                 * 具体原因是由于 cluster 模式 + 共享 UDP handle 导致的问题
                 *
                 * https://github.com/nodejs/node/blob/master/lib/internal/cluster/master.js#L282
                 */
                exclusive: true }, this.bindOptions));
            socket.on('error', consts_1.noop);
            socket.on('listening', onListening);
            socket.on('error', onError);
            let timeoutId = null;
            if (this.connectTimeout) {
                timeoutId = setTimeout(onTimeout, this.connectTimeout);
            }
            const self = this;
            function cleanup() {
                socket.removeListener('connect', onListening);
                socket.removeListener('error', onError);
                if (timeoutId) {
                    clearTimeout(timeoutId);
                }
            }
            function onListening() {
                socket.once('error', () => (socket.closed = true));
                socket.once('close', () => (socket.closed = true));
                if (self.customCreate) {
                    self.customCreate(socket);
                }
                cleanup();
                resolve(socket);
            }
            function onError(err) {
                try {
                    socket.close();
                }
                catch (err) {
                    // do nothing
                }
                cleanup();
                reject(err);
            }
            function onTimeout() {
                try {
                    socket.close();
                }
                catch (err) {
                    // do nothing
                }
                const err = new Error((0, util_1.format)('connect ETIMEDOUT. Timeout: %dms. Options: %O. BindOptions: %O', self.connectTimeout, self.connectOptions, self.bindOptions));
                // @ts-ignore
                err.errorno = 'ETIMEDOUT';
                // @ts-ignore
                err.code = 'ETIMEDOUT';
                // @ts-ignore
                err.syscall = 'connect';
                cleanup();
                reject(err);
            }
        });
    }
    async destroy(socket) {
        try {
            socket.close();
        }
        catch (err) {
            // do nothing
        }
        socket.unref();
        if (this.customDestroy) {
            this.customDestroy(socket);
        }
    }
    validate(socket) {
        /**
         * Check whether the socket has a reason for destroy
         * 检查 socket 是否有销毁原因
         */
        if (socket.destroyReason) {
            return false;
        }
        /**
         * Check whether the socket is closed
         * 检查 socket 是否已关闭
         */
        if (socket.closed) {
            socket.destroyReason = 'socket already closed';
            return false;
        }
        // <!-- Socket recycling strategy -->
        /**
         * Check socket survival time
         * 检查 socket 存活时间
         */
        if (socket.createdAt &&
            now - socket.createdAt >= this.maxSocketSurvivalTime) {
            socket.destroyReason = `socket survival time is greate than ${this.maxSocketSurvivalTime}ms`;
            return false;
        }
        /**
         * Check socket available times
         * 检查 socket 使用年龄
         */
        if (socket.age && socket.age >= this.maxSocketAge) {
            socket.destroyReason = `socket use age is greate than ${this.maxSocketAge}`;
            return false;
        }
        // <!-- Socket recycling strategy -->
        /**
         * Custom verification rules
         * 自定义校验规则
         */
        if (this.customValidate) {
            const result = this.customValidate(socket);
            if (result !== true) {
                socket.destroyReason =
                    socket.destroyReason ||
                        "socket didn't pass custom validate";
            }
            return result;
        }
        return true;
    }
    reset(socket, stage) {
        if (socket.closed) {
            return socket;
        }
        if (stage === 'acquire') {
            socket.age++;
        }
        if (this.customReset) {
            return this.customReset(socket);
        }
        return socket;
    }
}
exports.UDPOperation = UDPOperation;
