// @universe-design/react/lib/context-menu/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.contextMenu = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactDom = _interopRequireDefault(require("react-dom"));
var _useComplementEvent = require("../hooks/useComplementEvent");
var _hooks = require("../config-provider/hooks");
var _util = require("../util");
var _hooks2 = require("../hooks");
var _configProvider = require("../config-provider");
var _DropdownMenuContext = require("../dropdown/DropdownMenuContext");
require("./style");
const ContextMenu = props => {
  const {
    position,
    onClose,
    overlay
  } = props;
  const ref = _react.default.useRef(null);
  const containerRef = _react.default.useRef(null);
  const cssScope = (0, _hooks.useCssScope)('context-menu');
  // 点击非menu区域后关闭
  (0, _useComplementEvent.useComplementEvent)('click', {
    listener: _react.default.useCallback(() => {
      onClose && onClose();
    }, [onClose]),
    elements: _react.default.useMemo(() => [ref], [])
  });
  (0, _hooks2.useIsomorphicLayoutEffect)(() => {
    if (containerRef.current && ref.current) {
      // const Rect;
      const [left, top] = position;
      const result = {
        left,
        top
      };
      const el = ref.current;
      const cel = containerRef.current;
      const height = el.clientHeight;
      const width = el.clientWidth;

      // console.log('el', { left, top, height, width });
      // console.log('cel', { height: cel.clientHeight, width: cel.clientWidth });

      if (height + top > cel.clientHeight) {
        result.top = result.top - height;
        if (result.top < 0) {
          result.top = 0;
        }
      }
      if (width + left > cel.clientWidth) {
        result.left = left - width;
        if (result.left < 0) {
          result.left = 0;
        }
      }
      ref.current.style.left = `${result.left}px`;
      ref.current.style.top = `${result.top}px`;
    }
  }, [position]);
  const overlayElement = _react.default.useMemo(() => (0, _util.isFunction)(overlay) ? overlay() : overlay, [overlay]);
  return (
    /*#__PURE__*/
    // 为了告诉 menu 这是一个 dropdown
    _react.default.createElement(_DropdownMenuContext.DropdownMenuContext.Provider, {
      value: _react.default.useMemo(() => ({}), [])
    }, /*#__PURE__*/_react.default.createElement("div", {
      ref: containerRef,
      className: cssScope('__wrap'),
      onContextMenu: e => e.stopPropagation()
    }, /*#__PURE__*/_react.default.createElement("div", {
      ref: ref,
      className: cssScope('__overlay'),
      onClick: onClose
    }, overlayElement)))
  );
};
const contextMenu = exports.contextMenu = {
  open: option => {
    const {
      menu,
      position,
      getPopupContainer,
      onClose
    } = option;

    // const uuid = `ud-confirm-instance-${new Date().getTime()}`;

    const portalEl = getPopupContainer ? getPopupContainer() : document.body;
    const container = document.createElement('div');
    portalEl.appendChild(container);
    _reactDom.default.render(/*#__PURE__*/_react.default.createElement(_configProvider.CrossRootConfigProvider, null, /*#__PURE__*/_react.default.createElement(ContextMenu, {
      overlay: menu,
      position: position,
      onClose: () => {
        const unmountResult = _reactDom.default.unmountComponentAtNode(container);
        if (unmountResult && container.parentNode) {
          container.parentNode.removeChild(container);
          onClose && onClose();
        }
      }
    })), container);
  }
};
