// @universe-design/react/lib/transfer/panel/Search.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TransferSearch = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _SearchOutlineOutlined = _interopRequireDefault(require("@universe-design/icons-react/SearchOutlineOutlined"));
var _clsx = _interopRequireDefault(require("clsx"));
var _TransferContext = require("../hooks/TransferContext");
var _input = require("../../input");
var _configProvider = require("../../config-provider");
var _sx = require("../sx");
var _forwardRef = require("../../util/forwardRef");
const TransferSearch = exports.TransferSearch = (0, _forwardRef.forwardHTMLRef)((props, ref) => {
  const {
    className,
    onChange: onChangeProp,
    ...inputProps
  } = props;
  const {
    searchValue,
    setSearchValue,
    cssScope
  } = (0, _TransferContext.useTransferContext)();
  const {
    global: {
      search
    }
  } = (0, _configProvider.useLocale)();
  const onChange = React.useCallback(e => {
    onChangeProp === null || onChangeProp === void 0 ? void 0 : onChangeProp(e);
    setSearchValue(e.target.value);
  }, [onChangeProp, setSearchValue]);
  return /*#__PURE__*/React.createElement(_input.Input, (0, _extends2.default)({
    className: (0, _clsx.default)(className, cssScope('__search')),
    prefix: /*#__PURE__*/React.createElement(_SearchOutlineOutlined.default, null),
    placeholder: search
  }, inputProps, {
    value: searchValue,
    onChange: onChange,
    sx: [_sx.sxSearch],
    ref: ref
  }));
});
