// @universe-design/react/lib/transfer/panel/SourcePanel.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TransferSourcePanelHeader = exports.TransferSourcePanel = exports.TransferSelectAll = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var React = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _TransferContext = require("../hooks/TransferContext");
var _Search = require("./Search");
var _checkbox = require("../../checkbox");
var _PanelContainer = require("./PanelContainer");
var _SourceList = require("./SourceList");
var _list = require("../../shared/components/list");
var _configProvider = require("../../config-provider");
var _forwardRef = require("../../util/forwardRef");
var _sx = require("../sx");
/**
 * @public
 * `TransferSelectAll` 的类型。
 */

const TransferSelectAll = exports.TransferSelectAll = (0, _forwardRef.forwardHTMLRef)((props, ref) => {
  /** Context */
  const {
    cssScope,
    toggleSelectAll,
    isAllSelected,
    value,
    filteredOptions
  } = (0, _TransferContext.useTransferContext)();

  /** Locale */
  const {
    selectAll
  } = (0, _configProvider.useLocaleReceiver)('transfer');
  if (filteredOptions.length === 0) {
    return null;
  }

  /** Props */
  const {
    label = (selected, total) => selectAll.f({
      selected: String(selected),
      total: String(total)
    }),
    className,
    ...htmlProps
  } = props;

  /** Checked status */
  let indeterminate = false;
  const checked = isAllSelected;
  if (!checked && value.length) {
    indeterminate = filteredOptions.some(op => value.includes(op.value));
  }
  return /*#__PURE__*/React.createElement("div", (0, _extends2.default)({
    ref: ref,
    className: (0, _clsx.default)(className, cssScope('__select-all'), (0, _stylex.default)(_sx.sxSelectAll))
  }, htmlProps), /*#__PURE__*/React.createElement(_list.ListItemLike, null, /*#__PURE__*/React.createElement(_checkbox.Checkbox, {
    indeterminate: indeterminate,
    checked: checked,
    onCheckChange: toggleSelectAll
  }, label(value.length, filteredOptions.length))));
});
const TransferSourcePanelHeader = exports.TransferSourcePanelHeader = (0, _forwardRef.forwardHTMLRef)((props, ref) => {
  /** Memo default children */
  const defaultChildren = React.useMemo(() => /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(_Search.TransferSearch, null), /*#__PURE__*/React.createElement(TransferSelectAll, null)), []);

  /** Props */
  const {
    children = defaultChildren,
    ...htmlProps
  } = props;
  return /*#__PURE__*/React.createElement(_PanelContainer.TransferPanelHeader, (0, _extends2.default)({}, htmlProps, {
    ref: ref
  }), children);
});
const TransferSourcePanel = exports.TransferSourcePanel = (0, _forwardRef.forwardHTMLRef)((props, ref) => {
  /** Memo default children */
  const defaultChildren = React.useMemo(() => /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(TransferSourcePanelHeader, null), /*#__PURE__*/React.createElement(_SourceList.TransferSourceList, null)), []);

  /** Props */
  const {
    children = defaultChildren,
    optionRenderer,
    emptyContent: noOptionContentProp,
    ...htmlProps
  } = props;

  /** Context */
  const {
    searchValue,
    cssScope
  } = (0, _TransferContext.useTransferContext)();

  /** Locale */
  const {
    noSearchResult,
    noOptions
  } = (0, _configProvider.useLocaleReceiver)('transfer');
  const emptyContent = noOptionContentProp !== null && noOptionContentProp !== void 0 ? noOptionContentProp : searchValue ? noSearchResult : noOptions;
  return /*#__PURE__*/React.createElement(_PanelContainer.PanelContainer, (0, _extends2.default)({}, htmlProps, {
    emptyContent: emptyContent,
    optionRenderer: optionRenderer,
    ref: ref,
    className: (0, _clsx.default)(cssScope('__source-panel'), props.className)
  }), children);
});
