// @universe-design/react/lib/transfer/panel/SourceList.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TransferSourceList = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _rcVirtualList = _interopRequireDefault(require("@universe-design/rc-virtual-list"));
var React = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _TransferContext = require("../hooks/TransferContext");
var _utils = require("../utils");
var _PanelContext = require("../hooks/PanelContext");
var _TransferListCaption = require("./TransferListCaption");
var _sx = require("../sx");
var _TransferSourceOption = require("./TransferSourceOption");
var _OptionProvider = require("./OptionProvider");
var _AutoSizer = require("../../shared/components/AutoSizer");
var _highlighter = require("../../shared/components/highlighter");
var _useScrollStatus = require("../../hooks/useScrollStatus");
var _function = require("../../util/function");
var _useRcVirtualListResizeObserver = require("../../hooks/useRcVirtualListResizeObserver");
var _forwardRef = require("../../util/forwardRef");
/**
 * 默认AutoSizeDefaultHeight为0，数据量大的时候会导致rc-virtual-list 计算所有item的高度，造成浏览器假死
 * 设置一个默认10行高度的值，渲染过程中 AutoSize 会重新计算正确的高度
 */
const autoSizeDefaultHeight = 32 * 10;
const TransferSourceList = exports.TransferSourceList = (0, _forwardRef.forwardHTMLRef)((props, ref) => {
  /** Context */
  const {
    cssScope,
    filteredOptions,
    groupWeakMap,
    group,
    searchValue
  } = (0, _TransferContext.useTransferContext)();
  const {
    emptyContent
  } = (0, _PanelContext.useTransferPanelContext)();

  /** Props */
  const {
    className,
    listRef,
    ...htmlProps
  } = props;

  /** Derived value */
  const showGroup = Array.isArray(group) ? group[0] : group;
  const flatFilterOptions = React.useMemo(() => (0, _utils.flatTransferGroups)(filteredOptions, groupWeakMap, showGroup), [filteredOptions, groupWeakMap, showGroup]);

  /** Scroll status */
  const autoSizerRef = React.useRef(null);
  const {
    overflow,
    checkOverflow
  } = (0, _useScrollStatus.useScrollStatus)();
  const {
    vListHolderRef
  } = (0, _useRcVirtualListResizeObserver.useRcVirtualListResizeObserver)({
    getContainer: () => autoSizerRef.current,
    onHolderResize: () => {
      checkOverflow(vListHolderRef.current);
    },
    onHolderChildResize: () => {
      checkOverflow(vListHolderRef.current);
    }
  });

  /** List node */
  let listNode;
  if (flatFilterOptions.length === 0) {
    /** Empty */
    listNode = /*#__PURE__*/React.createElement("div", (0, _extends2.default)({}, htmlProps, {
      className: (0, _clsx.default)(cssScope('__panel__empty'), className, (0, _stylex.default)(_sx.sxPanelEmpty))
    }), emptyContent);
  } else {
    /** VList */
    listNode = /*#__PURE__*/React.createElement(_AutoSizer.AutoSizer, {
      className: (0, _stylex.default)(_sx.sxPanelListAutoSizer),
      ref: autoSizerRef,
      defaultHeight: autoSizeDefaultHeight,
      disableWidth: true
    }, _ref => {
      let {
        height
      } = _ref;
      return /*#__PURE__*/React.createElement(_rcVirtualList.default, {
        className: (0, _clsx.default)(cssScope('__panel__list'), overflow.t && [cssScope('__panel__list--scrolling'), (0, _stylex.default)(_sx.sxPanelListScrolling)]),
        itemKey: item => item.type === _utils.TransferListType.CAPION ? `group_key_${item.group.key}` : item.option.value,
        data: flatFilterOptions,
        ref: listRef,
        virtual: true,
        height: height,
        fullHeight: true,
        itemHeight: 32,
        onScroll: (0, _function.composeEventHandlers)(props.onScroll, ev => {
          checkOverflow(ev.currentTarget);
        })
      }, item => {
        if (item.type === _utils.TransferListType.CAPION) {
          return /*#__PURE__*/React.createElement(_TransferListCaption.TransferListCaption, null, item.group.title);
        }
        return (
          /*#__PURE__*/
          // item 必须有 ref
          React.createElement("div", null, /*#__PURE__*/React.createElement(_OptionProvider.TransferOptionProvider, {
            option: item.option
          }, /*#__PURE__*/React.createElement(_TransferSourceOption.TransferSourceOption, null)))
        );
      });
    });
  }

  /** List */
  return /*#__PURE__*/React.createElement(_highlighter.HighlightProvider, {
    value: {
      searchWords: [searchValue],
      disabled: false
    }
  }, /*#__PURE__*/React.createElement("div", (0, _extends2.default)({}, htmlProps, {
    className: (0, _clsx.default)(cssScope('__panel__list__wrapper'), className, (0, _stylex.default)(_sx.sxPanelListWrapper)),
    ref: ref
  }), listNode));
});
