// @universe-design/react/lib/transfer/panel/TargetPanel.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TransferTargetPanelHeader = exports.TransferTargetPanel = exports.TranferSelectionSummary = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var React = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _button = require("../../button");
var _TransferContext = require("../hooks/TransferContext");
var _PanelContainer = require("./PanelContainer");
var _TargetList = require("./TargetList");
var _list = require("../../shared/components/list");
var _configProvider = require("../../config-provider");
var _forwardRef = require("../../util/forwardRef");
var _sx = require("../sx");
const TranferSelectionSummary = exports.TranferSelectionSummary = (0, _forwardRef.forwardHTMLRef)((props, ref) => {
  /** Locale */
  const {
    selected,
    clear
  } = (0, _configProvider.useLocaleReceiver)('transfer');

  /** Props */
  const {
    showClearButton = true,
    selectionSummary = selectedCount => selected.f({
      selected: String(selectedCount)
    }),
    ...htmlProps
  } = props;

  /** Context */
  const {
    cssScope,
    value,
    clearAll
  } = (0, _TransferContext.useTransferContext)();
  return /*#__PURE__*/React.createElement("div", (0, _extends2.default)({
    className: (0, _clsx.default)((0, _stylex.default)(_sx.sxSelectionSummary), cssScope('__selection-summary'))
  }, htmlProps, {
    ref: ref
  }), /*#__PURE__*/React.createElement(_list.ListItemLike, null, selectionSummary(value.length)), showClearButton && value.length > 0 && /*#__PURE__*/React.createElement(_button.Button, {
    type: "text",
    onClick: clearAll,
    color: "primary",
    className: (0, _stylex.default)(_sx.sxClearButton)
  }, clear));
});
const TransferTargetPanelHeader = exports.TransferTargetPanelHeader = (0, _forwardRef.forwardHTMLRef)((props, ref) => {
  /** Memo default children */
  const defaultChildren = React.useMemo(() => /*#__PURE__*/React.createElement(TranferSelectionSummary, null), []);

  /** Props */
  const {
    children = defaultChildren,
    ...htmlProps
  } = props;
  return /*#__PURE__*/React.createElement(_PanelContainer.TransferPanelHeader, (0, _extends2.default)({}, htmlProps, {
    ref: ref
  }), children);
});
const TransferTargetPanel = exports.TransferTargetPanel = (0, _forwardRef.forwardHTMLRef)((props, ref) => {
  /** Memo default children */
  const defaultChildren = React.useMemo(() => /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(TransferTargetPanelHeader, null), /*#__PURE__*/React.createElement(_TargetList.TransferTargetList, null)), []);

  /** Props */
  const {
    children = defaultChildren,
    optionRenderer,
    emptyContent: noOptionContentProp,
    ...htmlProps
  } = props;

  /** Context */
  const {
    cssScope
  } = (0, _TransferContext.useTransferContext)();

  /** Locale */
  const {
    noOptionsSelected
  } = (0, _configProvider.useLocaleReceiver)('transfer');
  return /*#__PURE__*/React.createElement(_PanelContainer.PanelContainer, (0, _extends2.default)({}, htmlProps, {
    emptyContent: noOptionsSelected !== null && noOptionsSelected !== void 0 ? noOptionsSelected : noOptionContentProp,
    optionRenderer: optionRenderer,
    ref: ref,
    className: (0, _clsx.default)(cssScope('__target-panel'), props.className)
  }), children);
});
