// @universe-design/react/lib/transfer/panel/TransferTargetOption.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TransferTargetOptionRemoveButton = exports.TransferTargetOptionLabel = exports.TransferTargetOption = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _CloseSmallOutlined = _interopRequireDefault(require("@universe-design/icons-react/CloseSmallOutlined"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _clsx = _interopRequireDefault(require("clsx"));
var React = _interopRequireWildcard(require("react"));
var _button = require("../../button");
var _list = require("../../shared/components/list");
var _forwardRef = require("../../util/forwardRef");
var _TransferContext = require("../hooks/TransferContext");
var _TransferOptionContext = require("../hooks/TransferOptionContext");
var _sx = require("../sx");
var _utils = require("../utils");
const TransferTargetOptionRemoveButton = exports.TransferTargetOptionRemoveButton = /*#__PURE__*/React.forwardRef((props, ref) => {
  const {
    option,
    toggle,
    disabled
  } = (0, _TransferOptionContext.useTransferOptionContext)();
  const {
    cssScope
  } = (0, _TransferContext.useTransferContext)();
  const prefix = cssScope('__targetList');
  return /*#__PURE__*/React.createElement(_button.Button, (0, _extends2.default)({
    className: (0, _clsx.default)(`${prefix}-remove`, (0, _stylex.default)(_stylex.default.compose(_sx.sxRemoveButton, disabled && _sx.sxDisabledRemoveButton))),
    type: "icon",
    size: "small",
    icon: /*#__PURE__*/React.createElement(_CloseSmallOutlined.default, null),
    onClick: () => toggle(option.value),
    ref: ref
  }, props));
});
const TransferTargetOptionLabel = exports.TransferTargetOptionLabel = (0, _forwardRef.forwardHTMLRef)((props, ref) => {
  const {
    option
  } = (0, _TransferOptionContext.useTransferOptionContext)();
  return /*#__PURE__*/React.createElement(_list.ListItemTitle, (0, _extends2.default)({}, props, {
    ref: ref
  }), (0, _utils.optionLabel)(option));
});
const TransferTargetOption = exports.TransferTargetOption = (0, _forwardRef.forwardHTMLRef)((props, ref) => {
  var _props$sx;
  const defaultChildren = React.useMemo(() => /*#__PURE__*/React.createElement(TransferTargetOptionLabel, null), []);
  const defaultAfter = React.useMemo(() => /*#__PURE__*/React.createElement(TransferTargetOptionRemoveButton, null), []);
  const {
    children = defaultChildren,
    after = defaultAfter,
    ...restProps
  } = props;
  const {
    disabled
  } = (0, _TransferOptionContext.useTransferOptionContext)();
  return /*#__PURE__*/React.createElement(_list.ListItem, (0, _extends2.default)({
    disabled: disabled,
    after: after,
    ref: ref,
    showCheckmark: false
  }, restProps, {
    sx: [_sx.sxOption, ...((_props$sx = props.sx) !== null && _props$sx !== void 0 ? _props$sx : [])]
  }), children);
});
