// @universe-design/react/lib/transfer/panel/TransferSourceOption.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TransferSourceOptionLabel = exports.TransferSourceOptionCheckbox = exports.TransferSourceOption = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _clsx = _interopRequireDefault(require("clsx"));
var React = _interopRequireWildcard(require("react"));
var _checkbox = require("../../checkbox");
var _highlighter = require("../../shared/components/highlighter");
var _list = require("../../shared/components/list");
var _textOverflow = require("../../shared/components/text-overflow");
var _forwardRef = require("../../util/forwardRef");
var _function = require("../../util/function");
var _TransferContext = require("../hooks/TransferContext");
var _TransferOptionContext = require("../hooks/TransferOptionContext");
var _sx = require("../sx");
var _utils = require("../utils");
const TransferSourceOptionCheckbox = exports.TransferSourceOptionCheckbox = /*#__PURE__*/React.forwardRef((props, ref) => {
  const {
    isChecked,
    disabled
  } = (0, _TransferOptionContext.useTransferOptionContext)();
  return /*#__PURE__*/React.createElement(_checkbox.Checkbox, (0, _extends2.default)({}, props, {
    checked: isChecked,
    disabled: disabled,
    ref: ref
  }));
});
const TransferSourceOptionLabel = exports.TransferSourceOptionLabel = (0, _forwardRef.forwardHTMLRef)((props, ref) => {
  const {
    cssScope
  } = (0, _TransferContext.useTransferContext)();
  const {
    option,
    disabled
  } = (0, _TransferOptionContext.useTransferOptionContext)();
  const markRenderer = React.useCallback(text => /*#__PURE__*/React.createElement("span", {
    className: (0, _clsx.default)(!disabled && [cssScope('__label-highlight'), (0, _stylex.default)(_sx.sxLabelHighlight)])
  }, text), [cssScope, disabled]);
  return /*#__PURE__*/React.createElement(_list.ListItemTitle, (0, _extends2.default)({
    textOverflow: false
  }, props, {
    ref: ref
  }), /*#__PURE__*/React.createElement(_textOverflow.TextOverflow, {
    as: "div",
    className: (0, _clsx.default)((0, _stylex.default)(_sx.sxTransferLabel), cssScope('__source-option-label'), disabled && [(0, _stylex.default)(_sx.sxDisabledTransferCheckbox)]),
    lineClamp: 1
  }, /*#__PURE__*/React.createElement(_highlighter.HighlightConsumer, {
    markRenderer: markRenderer
  }, (0, _utils.optionLabel)(option))));
});
const TransferSourceOption = exports.TransferSourceOption = (0, _forwardRef.forwardHTMLRef)((props, ref) => {
  var _props$sx;
  const defaultChildren = React.useMemo(() => /*#__PURE__*/React.createElement(TransferSourceOptionLabel, null), []);
  const defaultBefore = React.useMemo(() => /*#__PURE__*/React.createElement(TransferSourceOptionCheckbox, null), []);
  const {
    children = defaultChildren,
    before = defaultBefore,
    ...restProps
  } = props;
  const {
    option,
    disabled,
    toggle
  } = (0, _TransferOptionContext.useTransferOptionContext)();
  (0, _function.composeEventHandlers)();
  return /*#__PURE__*/React.createElement(_list.ListItem, (0, _extends2.default)({
    showCheckmark: false,
    disabled: disabled,
    before: before,
    ref: ref
  }, restProps, {
    sx: [_sx.sxOption, ...((_props$sx = props.sx) !== null && _props$sx !== void 0 ? _props$sx : [])],
    onClick: React.useMemo(() => (0, _function.composeEventHandlers)(props.onClick, () => {
      if (disabled) {
        return;
      }
      toggle(option.value);
    }), [disabled, option.value, props.onClick, toggle])
  }), children);
});
