// @universe-design/react/es/anchor/AnchorLink.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React from 'react';
import clsx from 'clsx';
import stylex from '@universe-design/stylex';
import { cssTypographyHeadline, cssTypographyBody0 } from '../stylex';
import { isFunction } from '../util';
import { useCssScope } from '../config-provider/hooks';
import { useIsomorphicLayoutEffect as useLayoutEffect } from '../hooks';
import { useAnchorContext } from './AnchorContext';
import './style';

/**
 * Anchor.Link 组件的属性
 * @public
 */

export const AnchorLink = /*#__PURE__*/React.forwardRef((props, ref) => {
  const {
    href,
    title,
    target,
    children,
    className,
    onClick,
    ...restProps
  } = props;
  const cssScope = useCssScope('anchor-link');
  const {
    activeLink,
    registerLink,
    unregisterLink,
    handleLinkClick
  } = useAnchorContext();
  const linkRef = React.useRef(null);
  const cls = clsx(cssScope(''), className, {
    [cssScope('--active')]: activeLink === href
  });
  const titleCls = clsx(cssScope('__title'), {
    [cssScope('__title--active')]: activeLink === href
  }, stylex(cssTypographyBody0, activeLink === href && cssTypographyHeadline));
  useLayoutEffect(() => {
    registerLink(href, {
      ref: linkRef
    });
    return () => unregisterLink(href);
  }, [href, registerLink, unregisterLink]);
  return /*#__PURE__*/React.createElement("div", _extends({
    className: cls,
    ref: ref
  }, restProps), /*#__PURE__*/React.createElement("a", {
    ref: linkRef,
    className: titleCls,
    href: href,
    title: typeof title === 'string' ? title : '',
    target: target,
    onClick: e => {
      isFunction(onClick) && onClick(e, {
        title,
        href
      });
      handleLinkClick(e, {
        title,
        href
      });
    }
  }, title), children);
});
if (process.env.NODE_ENV !== "production") {
  AnchorLink.displayName = 'AnchorLink';
}
