// @universe-design/react/es/anchor/useAnchor.js
import React from 'react';
import clsx from 'clsx';
import { useControlledState, useConstant } from '@universe-design/react-utilities';
import { useCssScope, useGetTargetContainer } from '../config-provider/hooks';
import { isFunction, animationScrollTo, getSetScrollTop } from '../util';
import { useIsomorphicLayoutEffect as useLayoutEffect } from '../hooks';

/**
 * Anchor 组件的属性
 * @public
 */

function getOffsetTop(element, container) {
  if (!element.getClientRects().length) {
    return 0;
  }
  const rect = element.getBoundingClientRect();
  if (rect.width || rect.height) {
    if (container === window) {
      var _element$ownerDocumen, _element$ownerDocumen2, _element$ownerDocumen3;
      return rect.top - ((_element$ownerDocumen = (_element$ownerDocumen2 = element.ownerDocument) === null || _element$ownerDocumen2 === void 0 ? void 0 : (_element$ownerDocumen3 = _element$ownerDocumen2.documentElement) === null || _element$ownerDocumen3 === void 0 ? void 0 : _element$ownerDocumen3.clientTop) !== null && _element$ownerDocumen !== void 0 ? _element$ownerDocumen : 0);
    }
    return rect.top - container.getBoundingClientRect().top;
  }
  return rect.top;
}
function getElementByHash(hash) {
  const target = document.getElementById(hash.replace(/^#/, ''));
  return target;
}

/**
 * useAnchor
 * 名词解释
 * 1. anchorElement 锚定元素
 * 2. container 定位容器，所有的锚定元素都应该出现在定位容器中，因为我们需要通过滚动 container 来把锚定元素滚动到屏幕内部；此外，我们需要通过锚定元素和 container 的相对位置来计算当前激活的锚点
 *
 * 关键操作
 * 1. 注册 AnchorLink，根据 AnchorLink href 获取对应的锚定元素
 * 2. 注销 AnchorLink
 * 3. 监听所有锚定元素的 scrollParents，重新计算 activeLink
 * 4. 点击 AnchorLink，scroll 到对应的元素上去
 */
export const useAnchor = props => {
  const {
    affix = true,
    offsetTop = 0,
    targetOffset = offsetTop,
    bounds = 5,
    simple = false,
    showInk = true,
    getContainer: getContainerProp,
    onClick,
    onChange,
    className,
    ...restProps
  } = props;
  const cssScope = useCssScope('anchor');
  const ctxGetTargetContainer = useGetTargetContainer();
  const getContainer = getContainerProp !== null && getContainerProp !== void 0 ? getContainerProp : ctxGetTargetContainer;

  /**
   * Links
   */
  const [links, setLinks] = React.useState([]);
  const linkRegistry = useConstant(() => new Map());
  const registerLink = React.useCallback((link, payload) => {
    setLinks(preLink => {
      if (!preLink.includes(link)) {
        return [...preLink, link];
      }
      return preLink;
    });
    linkRegistry.set(link, payload);
  }, [linkRegistry]);
  const unregisterLink = React.useCallback(link => {
    setLinks(preLink => {
      if (preLink.includes(link)) {
        return preLink.filter(o => o !== link);
      }
      return preLink;
    });
    linkRegistry.delete(link);
  }, [linkRegistry]);

  /**
   * ActiveLink
   */
  const [activeLink, setActiveLink] = useControlledState(undefined, undefined, {
    onChange
  });
  const animatingRef = React.useRef(false);
  const forceUpdateActiveLink = React.useCallback(() => {
    if (animatingRef.current) {
      return;
    }
    const linkSections = [];
    const container = getContainer();
    links.forEach(link => {
      const target = getElementByHash(link);
      if (target) {
        const top = getOffsetTop(target, container);
        if (top < targetOffset + bounds) {
          linkSections.push({
            link,
            top
          });
        }
      }
    });
    if (linkSections.length) {
      const maxSection = linkSections.reduce((prev, curr) => curr.top > prev.top ? curr : prev);
      setActiveLink(maxSection.link);
      return;
    }
    setActiveLink(undefined);
  }, [bounds, getContainer, links, setActiveLink, targetOffset]);

  // 监听一下 scroll 时间，变更当前锚定元素
  React.useEffect(() => {
    const container = getContainer();
    if (!container) {
      return;
    }
    container.addEventListener('scroll', forceUpdateActiveLink);
    return () => container.removeEventListener('scroll', forceUpdateActiveLink);
  });
  const scrollToCleanupRef = React.useRef();
  const scrollToSection = React.useCallback(href => {
    // 调用 scrollTo 之前把上一个 scrollTo 先取消了
    if (scrollToCleanupRef.current) {
      scrollToCleanupRef.current();
    }
    const container = getContainer();
    if (!container) {
      return;
    }
    const scrollTop = getSetScrollTop(container);
    const targetElement = getElementByHash(href);
    if (!targetElement) {
      return;
    }
    const eleOffsetTop = getOffsetTop(targetElement, container);
    animatingRef.current = true;
    scrollToCleanupRef.current = animationScrollTo(scrollTop + eleOffsetTop - targetOffset, {
      scrollEnd: () => {
        animatingRef.current = false;
      },
      getContainer
    });
    setActiveLink(href);
  }, [setActiveLink, getContainer, targetOffset]);
  const handleLinkClick = React.useCallback((e, link) => {
    isFunction(onClick) && onClick(e, link);
    if (e.isDefaultPrevented()) {
      return;
    }
    scrollToSection(link.href);
  }, [onClick, scrollToSection]);
  /**
   * Anchor Ink
   */
  const getAnchorStyle = React.useCallback(() => {
    const style = {};
    if (activeLink) {
      var _linkRegistry$get;
      const linkNode = (_linkRegistry$get = linkRegistry.get(activeLink)) === null || _linkRegistry$get === void 0 ? void 0 : _linkRegistry$get.ref.current;
      if (linkNode) {
        const top = linkNode.offsetTop;
        const height = linkNode.offsetHeight;
        style.top = `${top}px`;
        style.height = `${height}px`;
      }
    }
    return style;
  }, [activeLink, linkRegistry]);
  /**
   * ================================================================
   *
   * Effects
   *
   * ================================================================
   */
  // forceUpdateActiveLink 变化的时候更新一下当前锚定元素，第一次对 window 滚动的场景有效，浏览器会默认滚到 hash 对应的 section
  useLayoutEffect(() => {
    forceUpdateActiveLink();
  }, [forceUpdateActiveLink]);
  return {
    // States
    activeLink,
    // Actions
    registerLink,
    unregisterLink,
    handleLinkClick,
    // Props
    affix,
    simple,
    showInk,
    // others
    cssScope,
    rootPrefixCls: cssScope(),
    getAnchorBodyProps: () => ({
      className: clsx(cssScope(), className, {
        [cssScope('--simple')]: simple
      }),
      ...restProps
    }),
    getAnchorInkProps: () => ({
      className: cssScope('__ink'),
      style: getAnchorStyle()
    }),
    getAffixProps: () => ({
      target: getContainer,
      offsetTop
    }),
    // Ref methods
    scrollToSection,
    forceUpdateActiveLink
  };
};
