// @universe-design/react/src/color-picker/ColorPickerContext.ts
import React, { useCallback } from 'react';
import isEqual from 'lodash/isEqual';
import { R500 } from '@universe-design/styles/esm/all.light.atom';
import type { EqualFunc } from '@universe-design/react-utilities';
import { createContextHook, useControlledState } from '@universe-design/react-utilities';
import type { ColorParam, ColorPickerBlocksProps, Color, HSL, RGB, HSV, ColorPickerChangeEvent } from './type';
import { defaultShowDarkCheckmark, formatColor } from './util';
import { useCssScope } from '../config-provider/hooks';
import { isFunction, isNonNullable } from '../util';
import useSyncState from '../hooks/useSyncState';
import { TinyColor } from '@universe-design/colors';
import { createContext, useContext, useContextSelector } from 'use-context-selector';
import { defaultColor } from './const';

interface IColorPickerContext extends Required<Pick<ColorPickerBlocksProps, 'showDarkCheckmark'>> {
  /**
   * 用户输入的颜色入参
   */
  value: ColorParam | undefined;
  /**
   * 用于 UI 展示的颜色，由于我们提供了 `transformDisplayColor` 属性用于在特殊场景下矫正色偏，所有 UI 和 value 对应的颜色可能会不同
   */
  displayColor: Color | undefined;
  color: Color | undefined;
  setValue: (value: ColorParam | undefined, event: ColorPickerChangeEvent) => void;
  /**
   * 当前选中颜色对应的 hsla
   */
  hsl: HSL;
  /**
   * 当前选中颜色对应的 rgba
   */
  rgb: RGB;
  /**
   * 当前选中颜色对应的 hsv
   */
  hsv: HSV;
  /**
   * 当前选中颜色对应的 hex
   */
  hex: string;
  rootCssScope: (suffix?: string) => string;
  transformDisplayColor: (color: Color) => Color;
}

const isColorEqual: EqualFunc<ColorParam | undefined> = (pre, cur) => {
  if (pre === undefined && cur === undefined) {
    return true;
  } else if (pre === undefined || cur === undefined) {
    return false;
  } else {
    return new TinyColor(pre).equals(cur);
  }
};

const {
  useContext: useColorPickerContext,
  useContextSelector: useColorPickerContextSelector,
  Provider: ColorPickerProvider,
} = createContextHook<IColorPickerContext>('ColorPickerContext', {
  strict: true,
  customizedContextMethods: {
    createContext,
    useContext,
    useContextSelector,
  },
});

const defaultTransformDisplayColor: NonNullable<ColorPickerBlocksProps['transformDisplayColor']> = c => c.hex;

function useColorPicker(props: ColorPickerBlocksProps): IColorPickerContext {
  const {
    defaultValue = R500,
    value: valueProp,
    showDarkCheckmark = defaultShowDarkCheckmark,
    prefixCls,
    onChange,
    transformDisplayColor: transformDisplayColorProp = defaultTransformDisplayColor,
  } = props;
  const cssScope = useCssScope('color-picker', prefixCls);
  const [controlledValue, _setControlledValue] = useControlledState<ColorParam | undefined>(defaultValue, valueProp, {
    isEqual,
    isControlledOverride: 'value' in props,
  });

  const setControlledValue = useCallback(
    (color: ColorParam | undefined, event: ColorPickerChangeEvent) => {
      if (isNonNullable(color)) {
        const colorObj = formatColor(color);
        isFunction(onChange) && onChange(colorObj.hex, colorObj, event);
        return;
      }
      isFunction(onChange) && onChange(undefined, undefined, event);
      _setControlledValue(color);
    },
    [_setControlledValue, onChange],
  );

  const [syncValue, setSyncValue] = useSyncState(controlledValue, {
    isEqual: isColorEqual,
  });

  const setValue = React.useCallback<IColorPickerContext['setValue']>(
    (_value, event: ColorPickerChangeEvent) => {
      setSyncValue(_value);
      setControlledValue(_value, event);
    },
    [setControlledValue, setSyncValue],
  );

  const formattedColor = React.useMemo(
    () => (syncValue !== undefined ? formatColor(syncValue) : syncValue),
    [syncValue],
  );

  const transformDisplayColor = React.useCallback<IColorPickerContext['transformDisplayColor']>(
    color => formatColor(transformDisplayColorProp(color)),
    [transformDisplayColorProp],
  );

  const displayColor = React.useMemo(
    () => formattedColor && transformDisplayColor(formattedColor),
    [formattedColor, transformDisplayColor],
  );
  return {
    value: syncValue,
    setValue,
    showDarkCheckmark,
    rootCssScope: cssScope,
    color: formattedColor,
    hsl: (formattedColor ?? defaultColor).hsl,
    hsv: (formattedColor ?? defaultColor).hsv,
    rgb: (formattedColor ?? defaultColor).rgb,
    hex: (formattedColor ?? defaultColor).hex,
    displayColor,
    transformDisplayColor,
  };
}

export { useColorPickerContext, ColorPickerProvider, useColorPicker, useColorPickerContextSelector };
