// @universe-design/react/src/color-picker/type.ts
import type { TinyColor } from '@universe-design/colors';
import type { CSSProperties, PropsWithChildren } from 'react';

/**
 * @public
 */
export interface RGB {
  /**
   * 红色
   */
  r: number;
  /**
   * 绿色
   */
  g: number;
  /**
   * 蓝色
   */
  b: number;
  /**
   * 透明度
   */
  a: number;
}

export interface HSL {
  /**
   * 色相
   */
  h: number;
  /**
   * 饱和度
   */
  s: number;
  /**
   * 亮度
   */
  l: number;
  /**
   * 透明度
   */
  a: number;
}

export interface HSV {
  /**
   * 色调
   */
  h: number;
  /**
   * 饱和度
   */
  s: number;
  /**
   * 明度
   */
  v: number;
  /**
   * 透明度
   */
  a: number;
}

/**
 * @public
 */
export interface Color {
  /**
   * HSL 形式的色值
   */
  hsl: HSL;
  /**
   * RGB 形式的色值
   */
  rgb: RGB;
  /**
   * HSV 形式的色值
   */
  hsv: HSV;
  /**
   * hex 字符串形式的色值
   */
  hex: string;
  /**
   * rgb 字符串
   */
  rgbString: string;
  /**
   * TinyColor 形式的色值
   */
  color: TinyColor;
}

export type ColorParam = string | HSL | HSV | RGB;

export enum HrgbKey {
  HEX = 'hex',
  R = 'r',
  G = 'g',
  B = 'b',
}

export enum ColorPickerStep {
  inHome,
  inMoreColor,
}

/**
 * @public
 * 变动来源，如手动在输入框输入、调节了色调滑块等
 */
export enum ColorPickerChangeSource {
  /**
   * 输入框
   */
  INPUT = 'input',
  /**
   * 色块选择
   */
  COLOR_BLOCK = 'colorBlock',
  /**
   * 色调滑块调整
   */
  HUE = 'hue',
  /**
   * 饱和度调整
   */
  SATURATION = 'saturation',
}

/**
 * @public
 */
export type ColorPickerChangeEvent = {
  /**
   * 变动来源，如手动在输入框输入、调节了色调滑块等
   */
  source: ColorPickerChangeSource;
};

/**
 * `<PrimitiveColorPicker />` 和 `<PaletteColorPicker />` 组件的属性。只包含纯数据逻辑的 `ColorPicker` 组件，适用于自定义组合内容的场景。
 * @public
 */
export interface ColorPickerBaseProps extends PropsWithChildren<unknown> {
  /**
   * 选中的色值
   */
  value?: ColorParam;
  /**
   * 选中色值后的回调
   */
  onChange?: {
    (hexString: string, color: Color, event: ColorPickerChangeEvent): void;
    (hexString: undefined, color: undefined, event: ColorPickerChangeEvent): void;
  };
  /**
   * 默认选中的色值
   * @defaultValue `#f54a45`
   */
  defaultValue?: ColorParam;
  prefixCls?: string;
  /**
   * `className` 属性透传
   */
  className?: string;
  /**
   * `style` 属性透传
   */
  style?: CSSProperties;
  /**
   * 显示颜色转化函数，用于在特殊场景下调整色偏。入参当前颜色值，出参为需要显示的颜色值。
   */
  transformDisplayColor?: (color: Color) => ColorParam;
}

/**
 * @public
 */
export interface ColorPickerBlocksProps extends ColorPickerBaseProps {
  /**
   * 根据选中的颜色切换勾号的颜色，默认会通过 WCAG2 对比度规范进行设置
   */
  showDarkCheckmark?: (hexString: string, color: Color) => boolean;
}
