// @byted-service/bytedtrace/lib/status/consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CounterSpanReportQps = exports.CounterTxnReportBps = exports.CounterTxnReportQps = exports.ErrLowerBoundSamplerMethodNumOverflow = exports.ErrIncRootSampledQpsCounterFail = exports.ErrMqConsumerSpanTypeInvalid = exports.ErrEmitMetricsFail = exports.ErrCounterTimerEmitFail = exports.ErrSerializeFail = exports.ErrSerializeAppendFail = exports.ErrIncSpanReportQpsCounterFail = exports.ErrIncTxnReportBpsCounterFail = exports.ErrIncTxnReportQpsCounterFail = exports.ErrIncAndGetSamplingRateFail = exports.ErrProviderFlushFail = exports.ErrAgentProviderFlushFail = exports.ErrNewDomainSocketClientFail = exports.ErrTcpDialFail = exports.ErrUnixSocketConnectFail = exports.ErrResolveUnixAddrFail = exports.ErrLogSpanFail = exports.ErrPostTraceCtxInvalid = exports.ErrReportedQueueFull = exports.ErrAgentProviderRePushQueueFull = exports.ErrAgentProviderQueueFull = exports.ErrRecordChildFail = exports.ErrTxnSamplingOverflow = exports.ErrPostTraceOverflow = exports.ErrDebugTraceOverflow = exports.ErrMetricNumInSpanOverflow = exports.ErrEventNumInSpanOverflow = exports.ErrTxnTagOverflow = exports.ErrAppendSpanNumOverflow = exports.ErrTxnDataLenOverflow = exports.ErrSpanTagOverflow = exports.ErrPanic = exports.COUNTER_SPAN_DEGRADE = exports.ROOT_POST_TRACE_QPS = exports.ROOT_DEBUG_QPS = exports.ROOT_LOWER_BOUND_SAMPLED_QPS = exports.ROOT_SAMPLED_QPS = void 0;
exports.ROOT_SAMPLED_QPS = 'root.sampled.qps';
exports.ROOT_LOWER_BOUND_SAMPLED_QPS = 'root.lower_bound_sampled.qps';
exports.ROOT_DEBUG_QPS = 'root.debug.qps';
exports.ROOT_POST_TRACE_QPS = 'root.ptrace.qps';
exports.COUNTER_SPAN_DEGRADE = 'span.degrade';
// error
exports.ErrPanic = 'panic';
exports.ErrSpanTagOverflow = 'span.tag.overflow';
exports.ErrTxnDataLenOverflow = 'txn.data.len.overflow';
exports.ErrAppendSpanNumOverflow = 'append.span.num.overflow';
exports.ErrTxnTagOverflow = 'txn.tag.overflow';
exports.ErrEventNumInSpanOverflow = 'event.num.in.span.overflow';
exports.ErrMetricNumInSpanOverflow = 'metric.num.in.span.overflow';
exports.ErrDebugTraceOverflow = 'debug.trace.overflow';
exports.ErrPostTraceOverflow = 'post.trace.overflow';
exports.ErrTxnSamplingOverflow = 'txn.sampling.overflow';
exports.ErrRecordChildFail = 'record.child.fail';
exports.ErrAgentProviderQueueFull = 'agent.provider.queue.full';
exports.ErrAgentProviderRePushQueueFull = 'agent.provider.rePush.queue.full';
exports.ErrReportedQueueFull = 'reporter.queue.full';
exports.ErrPostTraceCtxInvalid = 'post.trace.ctx.invalid';
exports.ErrLogSpanFail = 'log.span.fail';
exports.ErrResolveUnixAddrFail = 'resolve.unix.addr.fail';
exports.ErrUnixSocketConnectFail = 'unix.socket.connect.fail';
exports.ErrTcpDialFail = 'tcp.dial.fail';
exports.ErrNewDomainSocketClientFail = 'new.domain.socket.client.fail';
exports.ErrAgentProviderFlushFail = 'agent.provider.flush.fail';
exports.ErrProviderFlushFail = 'provider.flush.fail';
exports.ErrIncAndGetSamplingRateFail = 'inc.and.get.sampling.rate.fail';
exports.ErrIncTxnReportQpsCounterFail = 'inc.txn.report.qps.counter.fail';
exports.ErrIncTxnReportBpsCounterFail = 'inc.txn.report.bps.counter.fail';
exports.ErrIncSpanReportQpsCounterFail = 'inc.span.report.qps.counter.fail';
exports.ErrSerializeAppendFail = 'serialize.append.fail';
exports.ErrSerializeFail = 'serialize.fail';
exports.ErrCounterTimerEmitFail = 'counter.timer.emit.fail';
exports.ErrEmitMetricsFail = 'emit.metrics.fail';
exports.ErrMqConsumerSpanTypeInvalid = 'mq.consumer.span.type.invalid';
exports.ErrIncRootSampledQpsCounterFail = 'inc.root.sampled.qps.fail';
exports.ErrLowerBoundSamplerMethodNumOverflow = 'lower_bound.method.overflow';
exports.CounterTxnReportQps = 'txn.report.qps';
exports.CounterTxnReportBps = 'txn.report.Bps';
exports.CounterSpanReportQps = 'span.report.qps';
