// @byted-service/bytedtrace/lib/status/counter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StatusCounter = void 0;
const tslib_1 = require("tslib");
const sampler = tslib_1.__importStar(require("../sampler"));
const consts = tslib_1.__importStar(require("../consts"));
class StatusCounter {
    constructor(logger, metricsClient, enableCountingRate) {
        //  *sampler.LimiterPack
        // In debug mode, there are no warning logs
        this.debugMode = consts.IS_DEBUG;
        this.logger = logger;
        this.metrics = metricsClient;
        this.logger = logger;
        this.enableCountingRate = enableCountingRate;
        const samplerLimiter = new sampler.SimpleLimiter();
        samplerLimiter.setMaxHitPerSecond(consts.MAX_HIT_PER_SECOND_FOR_STAT_COUNTER);
        this.limiterPack = new sampler.LimiterPack(consts.MAX_OPERATION_FOR_LIMITER, sampler.DEFAULT_EXPIRED_TIME_MS, samplerLimiter);
        if (enableCountingRate) {
            this.counter = new sampler.QpsCatcher(false, {
                maxOperationNum: 100,
                windowDuringMs: 100,
                windowNum: 10,
                windowExpireTimeMs: 60 * 100 * 1e6, // 窗口过期时间ms
            });
        }
        else {
            this.counter = new sampler.QpsCatcher(true);
        }
    }
    /**
     * SDK 内部发生错误时进行 metrics 打点
     *
     * Tracer internal error metrics.
     * 'forceLog': if true, this log will be emitted whether is in debug mode.
     *  Format: bytedtrace.sdk.internal.error.rate{type=*}.
     * @param forceLog
     * @param location
     * @param formate
     * @param logs
     */
    recordWarningWithRateLimit(forceLog, location, formate, ...logs) {
        if (this.limiterPack.pass(location)) {
            if (forceLog || this.debugMode) {
                this.logger.warn(`[${location}] ${formate}`, ...logs);
            }
            if (this.metrics) {
                this.metrics.emitRateCounter('internal.error.rate', 1, undefined, {
                    type: location,
                });
            }
        }
    }
    async inc(name, delta, enableEmitMetrics) {
        this.counter.inc(name, delta);
        if (enableEmitMetrics && this.metrics) {
            this.metrics.emitRateCounter('internal.op.rate', delta, undefined, {
                name,
            });
        }
    }
    getRate(name) {
        return this.counter.getQps(name);
    }
    close() {
        this.counter.close();
    }
}
exports.StatusCounter = StatusCounter;
