// @universe-design/react/es/dialog/hooks/RefDialog.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React from 'react';
import { isFunction } from '../../util';
import { Confirm } from '../Confirm';
import { Modal } from '../Modal';
const createRefDialog = _ref => {
  let {
    Comp,
    componentName
  } = _ref;
  const C = /*#__PURE__*/React.forwardRef((_ref2, ref) => {
    let {
      defaultConfig
    } = _ref2;
    const [visible, setVisible] = React.useState(true);
    const [innerConfig, setInnerConfig] = React.useState(defaultConfig !== null && defaultConfig !== void 0 ? defaultConfig : {});
    React.useImperativeHandle(ref, () => ({
      destroy: () => {
        setVisible(false);
      },
      update: newConfig => {
        setInnerConfig(isFunction(newConfig) ? newConfig : originConfig => ({
          ...originConfig,
          ...newConfig
        }));
      }
    }));
    return /*#__PURE__*/React.createElement(Comp, _extends({}, innerConfig, {
      visible,
      onClose: function () {
        setVisible(false);
      }
    }));
  });
  if (process.env.NODE_ENV !== "production") {
    C.displayName = componentName;
  }
  return C;
};
export const RefConfirm = /* #__PURE__ */createRefDialog({
  Comp: Confirm,
  componentName: 'RefConfirm'
});
export const RefModal = /* #__PURE__ */createRefDialog({
  Comp: Modal,
  componentName: 'RefModal'
});
