// @universe-design/react/es/dialog/hooks/useDialog.js
/* eslint-disable max-nested-callbacks */
/**
 * @file useDialog
 */
import React from 'react';
import { isFunction } from '../../util';
import { withError, withInfo, withSuccess, withWarn } from '../Confirm';
import { RefConfirm, RefModal } from './RefDialog';
/**
 * 命令式操作 react elements 的渲染
 */
const ReactElements = /*#__PURE__*/React.forwardRef((_, ref) => {
  const [elements, setElements] = React.useState([]);
  const addNode = React.useCallback(element => {
    setElements(originElements => {
      if (originElements.includes(element)) {
        return originElements;
      }
      return [...originElements, element];
    });
    return () => {
      setElements(originElements => {
        if (originElements.includes(element)) {
          return originElements.filter(ele => ele !== element);
        }
        return originElements;
      });
    };
  }, []);
  React.useImperativeHandle(ref, () => ({
    addNode
  }), [addNode]);
  return /*#__PURE__*/React.createElement(React.Fragment, null, elements);
});
const EMPTY_ACTION_ARRAY = [];
// 自增 id
let uid = 0;
/**
 * useDialog 可以生成能获取到 context 的 dialog 实例
 */
export function useDialog() {
  const reactElementsRef = React.useRef(null);
  // actionQueue 用于解决 elementsHolder 和 dialog 组件还未 render 时 open update destroy 方法不生效的问题，每次都把没执行成功的 task 放入队列，下次 render 之后延迟执行
  const [actionQueue, dispatch] = React.useReducer((state, action) => {
    switch (action.type) {
      case 'enqueue':
        {
          return [...state, action.payload];
        }
      case 'dequeue':
        {
          if (!state.includes(action.payload)) {
            return state;
          }
          return state.filter(o => o !== action.payload);
        }
      default:
        {
          throw new Error('Action type is not valid.');
        }
    }
  }, EMPTY_ACTION_ARRAY);
  React.useEffect(() => {
    if (!actionQueue.length) {
      return;
    }
    for (const action of actionQueue) {
      dispatch({
        type: 'dequeue',
        payload: action
      });
      action();
    }
  }, [actionQueue]);
  const openDialog = React.useCallback((_ref, RefComp) => {
    let {
      onAfterClose,
      ...props
    } = _ref;
    uid += 1;
    const dialogRef = /*#__PURE__*/React.createRef();
    const addNode = () => {
      var _reactElementsRef$cur;
      const unmount = (_reactElementsRef$cur = reactElementsRef.current) === null || _reactElementsRef$cur === void 0 ? void 0 : _reactElementsRef$cur.addNode(/*#__PURE__*/React.createElement(RefComp, {
        key: `dialog-${uid}`,
        ref: dialogRef,
        defaultConfig: {
          ...props,
          // after close unmount 节点
          onAfterClose: () => {
            isFunction(onAfterClose) && onAfterClose();
            unmount === null || unmount === void 0 ? void 0 : unmount();
          }
        }
      }));
    };
    if (reactElementsRef.current) {
      addNode();
    } else {
      dispatch({
        type: 'enqueue',
        payload: addNode
      });
    }
    const update = config => {
      if (dialogRef.current) {
        dialogRef.current.update(config);
      } else {
        dispatch({
          type: 'enqueue',
          payload: () => update(config)
        });
      }
    };
    const destroy = () => {
      if (dialogRef.current) {
        dialogRef.current.destroy();
      } else {
        dispatch({
          type: 'enqueue',
          payload: () => destroy()
        });
      }
    };
    return {
      destroy,
      update
    };
  }, []);
  const dialogInstance = React.useMemo(() => ({
    confirm: props => openDialog(props, RefConfirm),
    info: props => openDialog(withInfo(props), RefConfirm),
    success: props => openDialog(withSuccess(props), RefConfirm),
    error: props => openDialog(withError(props), RefConfirm),
    warn: props => openDialog(withWarn(props), RefConfirm),
    open: props => openDialog(props, RefModal)
  }), [openDialog]);
  const reactElements = React.useMemo(() => /*#__PURE__*/React.createElement(ReactElements, {
    ref: reactElementsRef
  }), []);
  return [dialogInstance, reactElements];
}
