// @universe-design/react/es/dropdown/Button.js
import React from 'react';
import { Button } from '../button';
import { Dropdown } from './Dropdown';
import MoreOutlined from '@universe-design/icons-react/MoreOutlined';
import { useCssScope } from '../config-provider/hooks';
import clsx from 'clsx';
import './style';

/**
 * @public
 * @docgen
 * <Dropdown.Button> 的属性和方法。
 */

export const DropdownButton = props => {
  const {
    type,
    disabled,
    htmlType,
    children,
    overlay,
    trigger,
    align,
    visible,
    defaultVisible,
    placement = 'bottomRight',
    href,
    icon = /*#__PURE__*/React.createElement(MoreOutlined, null),
    title,
    className,
    style,
    getPopupContainer,
    onVisibleChange,
    buttonsRender,
    onClick,
    size,
    ...restProps
  } = props;
  const cssScope = useCssScope('dropdown-button');
  const dropdownProps = {
    align,
    overlay,
    disabled,
    visible,
    defaultVisible,
    placement,
    trigger,
    onVisibleChange,
    getPopupContainer,
    ...restProps
  };
  const leftButton = /*#__PURE__*/React.createElement(Button, {
    size: size,
    type: type,
    disabled: disabled,
    onClick: onClick,
    htmlType: htmlType,
    href: href,
    title: title
  }, children);
  const rightButton = /*#__PURE__*/React.createElement(Button, {
    size: size,
    disabled: disabled,
    type: type,
    shape: "square",
    icon: icon
  });
  const [leftButtonToRender, rightButtonToRender] = buttonsRender ? buttonsRender([leftButton, rightButton]) : [leftButton, rightButton];
  return /*#__PURE__*/React.createElement("div", {
    style: style,
    className: clsx(cssScope(), className)
  }, leftButtonToRender, /*#__PURE__*/React.createElement(Dropdown, dropdownProps, rightButtonToRender));
};
