// @universe-design/react/es/dropdown/Dropdown.js
import _extends from "@babel/runtime/helpers/esm/extends";
/* eslint-disable max-lines-per-function */

import React, { forwardRef, useCallback, useMemo, useRef } from 'react';
import { useCssScope } from '../config-provider/hooks';
import { usePopperSidecar } from '../config-provider/hooks/usePopperSidecar';
import { isFunction } from '../util';
import { DropdownMenuContext } from './DropdownMenuContext';
import composeRefs from '../util/composeRefs';
import { SlideUpDownTransition } from '../shared/components/transition/SlideUpDownTransition';
import './style';

/**
 * @public
 */

/**
 * @public
 * @docgen
 */

// eslint-disable-next-line max-lines-per-function
export const Dropdown = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    prefixCls: prefixClsInProps,
    align: alignInProps,
    placement = 'bottomLeft',
    showAction,
    hideAction,
    overlayClassName,
    overlayStyle,
    trigger = 'hover',
    disabled = false,
    onOverlayClick,
    overlay,
    children,
    udTriggerRef: udTriggerRefProp,
    ...otherProps
  } = props;
  const cssScope = useCssScope('dropdown');
  const Popper = usePopperSidecar();
  const prefixCls = prefixClsInProps || cssScope();
  const udTriggerRef = useRef(null);
  const mergedUdTriggerRef = React.useMemo(() => composeRefs(udTriggerRef, udTriggerRefProp), [udTriggerRef, udTriggerRefProp]);
  const overlayNode = useMemo(() => isFunction(overlay) ? overlay() : overlay, [overlay]);
  const onClick = useCallback(e => {
    var _udTriggerRef$current;
    isFunction(onOverlayClick) && onOverlayClick(e);
    if (e && 'domEvent' in e && e.domEvent.isDefaultPrevented()) {
      return;
    }
    (_udTriggerRef$current = udTriggerRef.current) === null || _udTriggerRef$current === void 0 ? void 0 : _udTriggerRef$current.setVisible(false);
  }, [onOverlayClick]);
  const renderOverlay = /*#__PURE__*/React.createElement(DropdownMenuContext.Provider, {
    value: useMemo(() => ({
      className: `${prefixCls}__overlay`,
      onClick
    }), [onClick, prefixCls])
  }, overlayNode);
  const alignPoint = trigger.indexOf('contextMenu') !== -1;
  return /*#__PURE__*/React.createElement(Popper, _extends({}, otherProps, {
    disabled: disabled,
    alignPoint: alignPoint,
    prefixCls: prefixCls,
    ref: ref,
    udTriggerRef: mergedUdTriggerRef,
    popupClassName: overlayClassName,
    popupStyle: overlayStyle,
    action: trigger,
    showAction: showAction,
    hideAction: hideAction,
    popupPlacement: placement,
    popupAlign: alignInProps,
    popupMotion: "slide-up",
    popup: renderOverlay,
    showArrow: false,
    arrowHasStroke: false,
    Transition: SlideUpDownTransition,
    defaultMainAxisOffsetWithoutArrow: 4,
    __screenshotAlwaysOpenEnabled: true
  }), React.useCallback(_ref => {
    let {
      visible
    } = _ref;
    return 'selected' in children.props ? children : /*#__PURE__*/React.cloneElement(children, {
      selected: visible
    });
  }, [children]));
});
if (process.env.NODE_ENV !== "production") {
  Dropdown.displayName = 'Dropdown';
}
