// @universe-design/react/es/upload/uploadList/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React from 'react';
import clsx from 'clsx';
import { UploadCardFile, UploadPictureFile, UploadSimpleTextFile, UploadTextFile } from '../uploadFile';
import { useUploadContext, useUploadContextSelector } from '../hooks/UploadContext';
import { Popover } from '../../popover';
import { Space } from '../../space';

// UploadCardFileList

const defaultCardItemRenderer = file => /*#__PURE__*/React.createElement(UploadCardFile, {
  file: file
});
export const UploadCardFileList = props => {
  const {
    itemRenderer = defaultCardItemRenderer,
    ...spaceProps
  } = props;
  const fileList = useUploadContextSelector(ctx => ctx.fileList);
  return /*#__PURE__*/React.createElement(Space, _extends({
    inline: false,
    direction: "horizontal"
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    ,
    size: [40, 16],
    wrap: true,
    align: "start"
  }, spaceProps, {
    itemKey: index => fileList[index].uid || index
  }), fileList.map(file => /*#__PURE__*/React.createElement(React.Fragment, {
    key: file.uid
  }, itemRenderer(file))));
};

// UploadTextFileList

const defaultTextItemRenderer = file => /*#__PURE__*/React.createElement(UploadTextFile, {
  file: file
});
export const UploadTextFileList = props => {
  const {
    itemRenderer = defaultTextItemRenderer,
    ...spaceProps
  } = props;
  const fileList = useUploadContextSelector(ctx => ctx.fileList);
  return /*#__PURE__*/React.createElement(Space, _extends({
    direction: "vertical",
    inline: false
  }, spaceProps, {
    itemKey: index => fileList[index].uid || index
  }), fileList.map(file => /*#__PURE__*/React.createElement(React.Fragment, {
    key: file.uid
  }, itemRenderer(file))));
};

// UploadSimpleTextFileList

const defaultSimpleTextItemRenderer = file => /*#__PURE__*/React.createElement(UploadSimpleTextFile, {
  file: file
});
export const UploadSimpleTextFileList = props => {
  const {
    itemRenderer = defaultSimpleTextItemRenderer,
    ...spaceProps
  } = props;
  const fileList = useUploadContextSelector(ctx => ctx.fileList);
  return /*#__PURE__*/React.createElement(Space, _extends({
    direction: "vertical",
    inline: false
  }, spaceProps, {
    itemKey: index => fileList[index].uid || index
  }), fileList.map(file => /*#__PURE__*/React.createElement(React.Fragment, {
    key: file.uid
  }, itemRenderer(file))));
};

// UploadPictureFileList
/**
 * @public
 */

export const UploadPictureFileList = props => {
  const {
    maxDisplayedPictureCount,
    uploadTrigger,
    itemRenderer = file => /*#__PURE__*/React.createElement(UploadPictureFile, {
      file: file
    }),
    ...spaceProps
  } = props;
  const {
    fileList,
    rootPrefixCls
  } = useUploadContext();
  let extraPictureList = [];
  let defaultDisplayedFiles = fileList;
  if (maxDisplayedPictureCount && fileList.length > maxDisplayedPictureCount) {
    extraPictureList = fileList.slice(maxDisplayedPictureCount - 1);
    defaultDisplayedFiles = fileList.slice(0, maxDisplayedPictureCount - 1);
  }
  const firstExtraFile = extraPictureList[0];
  const imgSrc = firstExtraFile ? firstExtraFile.thumbUrl || firstExtraFile.url : null;

  /**
   * space 默认会使用 index 作为 itemKey
   * 会使得 uploadTrigger 的 index 不断增加。导致 key 变化引发节点销毁，
   * 进而上传动作被取消
   */
  const itemKey = index => {
    if (index < defaultDisplayedFiles.length) {
      return String(defaultDisplayedFiles[index].uid);
    }
    if (extraPictureList.length && index === defaultDisplayedFiles.length) {
      return String(extraPictureList[0].uid);
    }
    return 'upload-trigger';
  };
  return (
    /*#__PURE__*/
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    React.createElement(Space, _extends({
      wrap: true,
      inline: false,
      size: [12, 6],
      itemKey: itemKey
    }, spaceProps), defaultDisplayedFiles.map(file => /*#__PURE__*/React.createElement(React.Fragment, {
      key: file.uid
    }, itemRenderer(file))), extraPictureList.length > 0 && /*#__PURE__*/React.createElement(Popover, {
      destroyOnHide: false,
      placement: "bottom",
      content: /*#__PURE__*/React.createElement("div", {
        className: `${rootPrefixCls}-file-list ${rootPrefixCls}-item--picture__extra__list`
      }, extraPictureList.map(_file => /*#__PURE__*/React.createElement(React.Fragment, {
        key: _file.uid
      }, itemRenderer(_file))))
    }, /*#__PURE__*/React.createElement("div", {
      className: clsx(`${rootPrefixCls}-item`, `${rootPrefixCls}-item--picture`, `${rootPrefixCls}-item--picture__extra`)
    }, imgSrc && /*#__PURE__*/React.createElement("img", {
      src: imgSrc
    }), /*#__PURE__*/React.createElement("div", {
      className: `${rootPrefixCls}-item--picture-mask`
    }), /*#__PURE__*/React.createElement("div", {
      className: `${rootPrefixCls}-item--picture__extra-count`
    }, "+", extraPictureList.length))), uploadTrigger)
  );
};
