// @universe-design/react/es/upload/uploadList/PictureUploadTrigger.js
import _extends from "@babel/runtime/helpers/esm/extends";
import AddOutlined from '@universe-design/icons-react/AddOutlined';
import clsx from 'clsx';
import React from 'react';
import { forwardHTMLRef } from '../../util/forwardRef';
import { useUploadContext } from '../hooks/UploadContext';
import { UploadTrigger } from '../UploadTrigger';
export const PictureUploadTrigger = forwardHTMLRef((props, ref) => {
  const {
    accept,
    multiple,
    directory,
    type,
    openFileDialogOnClick,
    ...restProps
  } = props;
  const {
    disabled,
    rootPrefixCls
  } = useUploadContext();
  return /*#__PURE__*/React.createElement(UploadTrigger, {
    accept: accept,
    multiple: multiple,
    directory: directory,
    type: type,
    openFileDialogOnClick: openFileDialogOnClick
  }, /*#__PURE__*/React.createElement("div", _extends({
    className: clsx({
      [`${rootPrefixCls}-picture-upload-trigger`]: true,
      [`${rootPrefixCls}-picture-upload-trigger--disabled`]: disabled,
      [`${rootPrefixCls}-picture-upload-trigger--active`]: !disabled
    }),
    ref: ref
  }, restProps), /*#__PURE__*/React.createElement(AddOutlined, null)));
});
