// @universe-design/react/src/input/stories/homeless.stories.tsx
import React, { useState } from 'react';
import type { InputProps } from '..';
import { Input } from '..';
import EmojiFilled from '@universe-design/icons-react/EmojiFilled';
import ContactsOutlined from '@universe-design/icons-react/ContactsOutlined';
import InfoOutlined from '@universe-design/icons-react/InfoOutlined';
import EmojiOutlined from '@universe-design/icons-react/EmojiOutlined';
import type { DefaultSizeType, TagProps } from '../..';
import {
  Tooltip,
  Col,
  Row,
  Select,
  InputNumber,
  Cascader,
  TreeSelect,
  TreeNode,
  Tag,
  DatePicker,
  Radio,
  Button,
  Divider,
  Switch,
  Space,
  Form,
} from '../..';
import type { Searchhape } from '../Search';

export default {
  title: 'Components/Input/临时示例', // 注意：这里需要修改成自己对应 Component name。
  component: Input,
};

export * from './textarea.stories';

const renderSelecter = (size: DefaultSizeType) => (
  <div onMouseDown={e => e.stopPropagation()}>
    <Select style={{ width: 100 }} allowClear={false} size={size} bordered={false}>
      <Select.Option value=".cn">.cn</Select.Option>
      <Select.Option value=".com">.com</Select.Option>
    </Select>
  </div>
);

const getTagSize = (s: DefaultSizeType) => {
  if (s === 'small') {
    return 'extra-small';
  }
  if (s === 'medium') {
    return 'small';
  }
  return 'medium';
};

export const Controlled = () => {
  const [value, setValue] = useState<string>();
  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    console.log('value change, value is', e.target.value);
    setValue(e.target.value);
  };
  return <Input allowClear value={value} onChange={onChange} />;
};

export const AllInput = () => {
  const [size, setSize] = useState<DefaultSizeType>('medium');
  const [disabled, setDisabled] = useState<boolean>(false);
  const [error, setError] = useState<boolean>(false);
  return (
    <div>
      <div>
        <span>尺寸：</span>
        <Radio.Group defaultValue="medium" onChange={val => setSize(val as DefaultSizeType)}>
          <Radio.Button value="large">大</Radio.Button>
          <Radio.Button value="medium">中</Radio.Button>
          <Radio.Button value="small">小</Radio.Button>
        </Radio.Group>
        <span style={{ marginLeft: 20 }}>禁用：</span>
        <Switch checked={disabled} onChange={setDisabled} />
        <span style={{ marginLeft: 30 }}>错误：</span>
        <Switch checked={error} onChange={setError} />
      </div>
      <br />
      <br />
      <Input size={size} disabled={disabled} error={error} placeholder="small size" prefix={<EmojiFilled />} />
      <br />
      <br />
      <Input size={size} disabled={disabled} error={error} placeholder="small size" addonAfter={<EmojiFilled />} />
      <br />
      <br />
      <Input
        size={size}
        prefix={<ContactsOutlined />}
        placeholder="这是一段超长的 placeholder。这是一段超长的 placeholder。"
        suffix={
          <Tooltip placement="top" title="Extra information">
            <Button type="icon" icon={<InfoOutlined />} />
          </Tooltip>
        }
        allowClear
        disabled={disabled}
        error={error}
      />
      <br />
      <br />
      <Input
        size={size}
        disabled={disabled}
        error={error}
        placeholder="请输入文字"
        suffix={
          <Tag disabled={disabled} color="neutral-option" size={getTagSize(size)}>
            English
          </Tag>
        }
      />
      <br />
      <br />
      <Input size={size} disabled={disabled} error={error} autoFocus placeholder="请输入文字" addonAfter="%" />
      <br />
      <br />
      <Input size={size} disabled={disabled} error={error} placeholder="请输入文字" addonBefore="http://" />
      <br />
      <br />
      <Input.Password size={size} disabled={disabled} error={error} placeholder="请输入密码" />
      <br />
      <br />
      <Input
        size={size}
        disabled={disabled}
        error={error}
        prefix={<Button type="icon" size={size} icon={<EmojiOutlined />} />}
        suffix={<Button type="icon" size={size} icon={<EmojiOutlined />} />}
      />
      <br />
      <br />
      <Input
        size={size}
        disabled={disabled}
        error={error}
        suffix={<Button type="icon" size={size} icon={<EmojiOutlined />} />}
        placeholder="请输入文字"
      />
      <br />
      <br />
      <Input
        size={size}
        disabled={disabled}
        error={error}
        addonBefore={
          <Select size={size} style={{ width: 100 }}>
            <Select.Option value="http://">http://</Select.Option>
            <Select.Option value="https://">https://</Select.Option>
          </Select>
        }
        addonAfter={
          <Select size={size} style={{ width: 100 }}>
            <Select.Option value=".cn">.cn</Select.Option>
            <Select.Option value=".com">.com</Select.Option>
          </Select>
        }
      />
      <br />
      <br />
      <Input
        size={size}
        disabled={disabled}
        error={error}
        prefix={renderSelecter(size)}
        suffix={renderSelecter(size)}
      />
      <br />
      <br />
      <Input
        size={size}
        disabled={disabled}
        error={error}
        placeholder="small size"
        suffix=".com"
        allowClear
        addonAfter={<EmojiFilled />}
      />
      <br />
      <br />
      <Input
        size={size}
        disabled={disabled}
        error={error}
        placeholder="small size"
        prefix=".com"
        addonBefore={<EmojiFilled />}
      />
      <br />
      <br />
      <Input size={size} disabled={disabled} error={error} placeholder="small size" style={{ width: 90 }} />
      <br />
      <br />
      <Input size={size} disabled={disabled} error={error} allowClear placeholder="small size" />
    </div>
  );
};

export const TextArea = () => {
  const [value, setValue] = useState<string>('');
  const [disabled, setDisabled] = useState<boolean>(false);
  const [error, setError] = useState<boolean>(false);

  return (
    <div>
      <div>
        <span>禁用：</span>
        <Switch checked={disabled} onChange={setDisabled} />
      </div>
      <div>
        <span>错误：</span>
        <Switch checked={error} onChange={setError} />
      </div>
      <Divider text="基础" />
      <Input.TextArea allowClear placeholder="请输入" disabled={disabled} error={error} />
      <Input.TextArea allowClear placeholder="请输入" defaultValue="hello word" disabled={disabled} error={error} />
      <Divider text="展示数量" />
      <Input.TextArea allowClear placeholder="请输入" showCount disabled={disabled} error={error} />
      <Divider text="展示数量 - 设置最大长度" />
      <Input.TextArea
        style={{ width: 300 }}
        allowClear
        placeholder="请输入"
        showCount
        maxLength={100}
        disabled={disabled}
        error={error}
      />
      <Divider text="自适应高度" />
      <Input.TextArea allowClear autoSize placeholder="自适应高度" disabled={disabled} error={error} />
      <Divider text="自适应高度 - 设置最大最小边界（最小5行，最大10行）" />
      <Input.TextArea
        allowClear
        rows={1}
        autoSize={{ minRows: 5, maxRows: 10 }}
        placeholder="自适应高度"
        disabled={disabled}
        error={error}
      />
      <Divider text="受控" />
      <Input.TextArea
        allowClear
        placeholder="请输入"
        value={value}
        onChange={e => setValue(e.target.value)}
        disabled={disabled}
        error={error}
      />
    </div>
  );
};

export const SearchInput = () => {
  const [size, setSize] = useState<DefaultSizeType>('medium');
  const [shape, setShape] = useState<Searchhape>('round');
  const [error, setError] = useState<boolean>(false);

  return (
    <div>
      <div>
        <span>尺寸：</span>
        <Radio.Group defaultValue="medium" onChange={val => setSize(val as DefaultSizeType)}>
          <Radio.Button value="large">大</Radio.Button>
          <Radio.Button value="medium">中</Radio.Button>
          <Radio.Button value="small">小</Radio.Button>
        </Radio.Group>
        <span style={{ marginLeft: 30 }}>形状：</span>
        <Radio.Group defaultValue="round" onChange={val => setShape(val as Searchhape)}>
          <Radio.Button value="round">全圆角</Radio.Button>
          <Radio.Button value="square">矩形</Radio.Button>
        </Radio.Group>
        <span style={{ marginLeft: 30 }}>错误：</span>
        <Switch checked={error} onChange={setError} />
      </div>
      <Divider text="基础" />
      <Input.Search size={size} error={error} shape={shape} allowClear />
      <Divider text="自定义后缀" />
      <Input.Search size={size} error={error} shape={shape} allowClear suffix="(⌘+K)" />
    </div>
  );
};

export const Group = () => {
  const { Option } = Select;
  const [size, setSize] = useState<DefaultSizeType>('medium');
  const [error, setError] = useState<boolean>(false);

  const options = [
    {
      value: 'zhejiang',
      label: 'Zhejiang',
      children: [
        {
          value: 'hangzhou',
          label: 'Hangzhou',
          children: [
            {
              value: 'xihu',
              label: 'West Lake',
            },
          ],
        },
      ],
    },
    {
      value: 'jiangsu',
      label: 'Jiangsu',
      children: [
        {
          value: 'nanjing',
          label: 'Nanjing',
          children: [
            {
              value: 'zhonghuamen',
              label: 'Zhong Hua Men',
            },
          ],
        },
      ],
    },
  ];
  return (
    <Space className="site-input-group-wrapper" direction="vertical">
      <div>
        <span>尺寸：</span>
        <Radio.Group defaultValue="medium" onChange={val => setSize(val as DefaultSizeType)}>
          <Radio.Button value="large">大</Radio.Button>
          <Radio.Button value="medium">中</Radio.Button>
          <Radio.Button value="small">小</Radio.Button>
        </Radio.Group>
        <span style={{ marginLeft: 30 }}>错误：</span>
        <Switch checked={error} onChange={setError} />
      </div>
      <br />
      <br />
      <Input.Group size={size}>
        <Row gutter={8}>
          <Col span={5}>
            <Input defaultValue="0571" error={error} />
          </Col>
          <Col span={8}>
            <Input defaultValue="26888888" error={error} />
          </Col>
        </Row>
      </Input.Group>
      <br />
      <Input.Group compact size={size}>
        <Select style={{ width: 150 }} defaultValue="https" error={error} allowClear={false}>
          <Option value="http:">http://</Option>
          <Option value="https">https://</Option>
        </Select>
        <Input style={{ width: '50%' }} defaultValue="www.baidu.com" error={error} addonAfter=".com" />
      </Input.Group>
      <br />
      <Input.Group compact size={size}>
        <Input style={{ width: '20%' }} defaultValue="0571" error={error} />
        <Input style={{ width: '30%' }} defaultValue="26888888" error={error} />
      </Input.Group>
      <br />
      <Input.Group compact size={size}>
        <Input style={{ width: '20%' }} defaultValue="0571" error={error} />
        <Input style={{ width: '30%' }} defaultValue="26888888" error={error} />
        <Input style={{ width: '30%' }} defaultValue="26888888" error={error} />
      </Input.Group>
      <br />
      <Input.Group compact size={size}>
        <Select style={{ width: 150 }} defaultValue="Zhejiang" error={error} allowClear={false}>
          <Option value="Zhejiang">Zhejiang</Option>
          <Option value="Jiangsu">Jiangsu</Option>
        </Select>
        <Input style={{ width: '50%' }} defaultValue="Xihu District, Hangzhou" error={error} />
      </Input.Group>
      <br />
      <Input.Group compact size={size}>
        <Input style={{ width: '20%' }} defaultValue="input content" error={error} />
        <DatePicker style={{ width: '50%' }} error={error} />
      </Input.Group>
      <br />
      <Input.Group compact size={size}>
        <DatePicker style={{ width: '50%' }} error={error} />
        <Input style={{ width: '20%' }} defaultValue="input content" error={error} />
      </Input.Group>
      <br />
      <Input.Group compact size={size}>
        <Input style={{ width: '20%' }} defaultValue="input content" error={error} />
        <DatePicker style={{ width: '50%' }} error={error} />
        <Input style={{ width: '20%' }} defaultValue="input content" error={error} />
      </Input.Group>
      <br />
      <Input.Group compact size={size}>
        <Input style={{ width: '20%' }} defaultValue="input content" error={error} />
        <DatePicker.RangePicker style={{ width: '50%' }} error={error} />
      </Input.Group>
      <br />
      <Input.Group compact size={size}>
        <DatePicker.RangePicker style={{ width: '50%' }} error={error} />
        <Input style={{ width: '20%' }} defaultValue="input content" error={error} />
      </Input.Group>
      <br />
      <Input.Group compact size={size}>
        <Input style={{ width: '20%' }} defaultValue="input content" error={error} />
        <DatePicker.RangePicker style={{ width: '50%' }} error={error} />
        <Input style={{ width: '20%' }} defaultValue="input content" error={error} />
      </Input.Group>
      <br />
      <Input.Group compact size={size}>
        <Input.Search shape="square" style={{ width: '40%' }} defaultValue="0571" error={error} />
        <Input.Search shape="square" allowClear style={{ width: '40%' }} defaultValue="26888888" error={error} />
      </Input.Group>
      <br />
      <Input.Group compact size={size}>
        <Input.Search shape="square" allowClear style={{ width: '40%' }} defaultValue="0571" error={error} />
        <Input.Search shape="square" allowClear style={{ width: '40%' }} defaultValue="26888888" error={error} />
      </Input.Group>
      <br />
      <Input.Group compact size={size}>
        <Select style={{ width: '30%' }} defaultValue="Option1" error={error}>
          <Option value="Option1">Option1</Option>
          <Option value="Option2">Option2</Option>
        </Select>
        <Input style={{ width: '50%' }} defaultValue="input content" error={error} />
        <InputNumber error={error} style={{ width: '20%' }} />
      </Input.Group>
      <br />
      <Input.Group compact size={size}>
        <InputNumber error={error} style={{ width: '20%' }} />
        <Select style={{ width: '30%' }} defaultValue="Option1" error={error}>
          <Option value="Option1">Option1</Option>
          <Option value="Option2">Option2</Option>
        </Select>
        <Input style={{ width: '50%' }} defaultValue="input content" error={error} />
      </Input.Group>
      <br />
      <Input.Group compact size={size}>
        <Select style={{ width: '30%' }} defaultValue="Option1" error={error}>
          <Option value="Option1">Option1</Option>
          <Option value="Option2">Option2</Option>
        </Select>
        <InputNumber style={{ width: '20%' }} error={error} />
        <Input style={{ width: '50%' }} defaultValue="input content" error={error} />
      </Input.Group>
      <br />
      <Input.Group compact size={size}>
        <Select style={{ width: 150 }} defaultValue="Option1-1" error={error} allowClear={false}>
          <Option value="Option1-1">Option1-1</Option>
          <Option value="Option1-2">Option1-2</Option>
        </Select>
        <Select style={{ width: 150 }} defaultValue="Option2-2" error={error} allowClear={false}>
          <Option value="Option2-1">Option2-1</Option>
          <Option value="Option2-2">Option2-2</Option>
        </Select>
      </Input.Group>
      <br />
      <Input.Group compact size={size}>
        <Select style={{ width: 150 }} defaultValue="1" error={error} allowClear={false}>
          <Option value="1">Between</Option>
          <Option value="2">Except</Option>
        </Select>
        <Input style={{ width: 100, textAlign: 'center' }} placeholder="Minimum" error={error} />
        <Input
          className="site-input-split"
          style={{
            width: 30,
            borderRight: 0,
            pointerEvents: 'none',
          }}
          placeholder="~"
          disabled
        />
        <Input
          className="site-input-right"
          style={{
            width: 100,
            textAlign: 'center',
          }}
          placeholder="Maximum"
          error={error}
        />
      </Input.Group>
      <br />
      <Input.Group compact size={size}>
        <Select style={{ width: '30%' }} defaultValue="Home" error={error} allowClear={false}>
          <Option value="Home">Home</Option>
          <Option value="Company">Company</Option>
        </Select>
        <Cascader style={{ width: '70%' }} options={options} placeholder="Select Address" error={error} />
      </Input.Group>
      <br />
      <Input.Group compact size={size}>
        <Cascader style={{ width: '70%' }} options={options} placeholder="Select Address" error={error} />
        <Select style={{ width: '30%' }} defaultValue="Home" error={error} allowClear={false}>
          <Option value="Home">Home</Option>
          <Option value="Company">Company</Option>
        </Select>
      </Input.Group>
      <br />
      <Input.Group compact size={size}>
        <TreeSelect
          showSearch
          style={{ width: '70%' }}
          dropdownStyle={{ maxHeight: 400, overflow: 'auto' }}
          placeholder="Please select"
          allowClear
          treeDefaultExpandAll
          error={error}
        >
          <TreeNode value="parent 1" title="parent 1">
            <TreeNode value="parent 1-0" title="parent 1-0">
              <TreeNode value="leaf1" title="my leaf" />
              <TreeNode value="leaf2" title="your leaf" />
            </TreeNode>
            <TreeNode value="parent 1-1" title="parent 1-1">
              <TreeNode value="sss" title={<b style={{ color: '#08c' }}>sss</b>} />
            </TreeNode>
          </TreeNode>
        </TreeSelect>
        <Select style={{ width: '30%' }} defaultValue="Home" error={error} allowClear={false}>
          <Option value="Home">Home</Option>
          <Option value="Company">Company</Option>
        </Select>
      </Input.Group>
    </Space>
  );
};

export const TextAreaClear = () => (
  <Input.TextArea
    style={{ width: 200 }}
    // allowClear
    showCount
    // autoSize
    maxLength={100}
    autoSize={{ minRows: 2, maxRows: 4 }}
    placeholder="自适应高度"
  />
);

export const OnClear = () => {
  const [] = useState();
  const onClear = () => console.log('onclear!');
  return (
    <Space direction="vertical" size="large" style={{ width: '50%' }}>
      <Input allowClear onClear={onClear} />
      <Input.Password allowClear onClear={onClear} />
      <Input.Search allowClear onClear={onClear} />
      <Input.TextArea rows={5} allowClear onClear={onClear} />
    </Space>
  );
};

export const CounterWithSize = () => (
  <Space size="large" direction="vertical">
    {(['large', 'medium', 'small'] as InputProps['size'][]).map(size => {
      const tagSize: TagProps['size'] = size === 'large' ? 'medium' : 'small';
      return (
        <>
          <h4>Size: {size}</h4>
          <Input showCount={true} maxLength={10} size={size} placeholder="Input password" />
          <Input
            showCount={true}
            maxLength={10}
            counter={str => [...str].length}
            truncateExcessWords
            size={size}
            suffix={<ContactsOutlined />}
            allowClear
            placeholder="Input password"
          />
          <Input
            showCount={true}
            maxLength={10}
            truncateExcessWords
            size={size}
            suffix={
              <Tag size={tagSize} color="neutral-light">
                中文
              </Tag>
            }
            placeholder="Input password"
          />
        </>
      );
    })}
  </Space>
);

export const WithForm = () => {
  const [formValue, setFormValue] = useState<{
    a: string;
    b: string;
    c: string;
    d: string;
  }>({ a: '', b: '', c: '', d: '' });
  return (
    <>
      <p>测试受控值，及中文输入</p>
      <Form initialValues={formValue} onValuesChange={v => setFormValue({ ...formValue, ...v })}>
        <Form.Item name="a" label="最大长度 10">
          <Input maxLength={10} onChange={e => console.log(e.target.value)} />
        </Form.Item>
        <Form.Item name="b" label="Show Counter and truncate">
          <Input
            maxLength={10}
            showCount
            truncateExcessWords
            onCompositionEnd={() => console.log('composition ended')}
            onCompositionStart={() => console.log('composition started')}
            onChange={e => console.log('onChange: ', e.target.value)}
            allowClear
          />
        </Form.Item>
        <Form.Item name="c" label="Show Counter">
          <Input maxLength={10} showCount onChange={e => console.log(e.target.value)} allowClear />
        </Form.Item>
        <Form.Item name="d" label="Show Counter  and truncate">
          <Input.TextArea
            maxLength={10}
            showCount
            truncateExcessWords
            onChange={e => console.log(e.target.value)}
            allowClear
          />
        </Form.Item>
      </Form>
      <p>得到值： {JSON.stringify(formValue)}</p>
    </>
  );
};
