// @universe-design/react/src/input/stories/screenshot.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C999Case0 from '../demo/999.case0';
import CInputCounter from '../demo/InputCounter';
import CAutofocusScreenshot from '../demo/autofocus.screenshot';
import CDisabled from '../demo/disabled';
import CExtraSmallContainer from '../demo/extra-small-container';
import CSizeScreenshot from '../demo/size.screenshot';
import CSuffixAndPrefixScreenshot from '../demo/suffix-and-prefix.screenshot';
import CTextareaDisabled from '../demo/textarea-disabled';
import CTextareaHeight from '../demo/textarea-height';
if (!(C999Case0 as Story).storyName) {
  (C999Case0 as Story).storyName = '999 case0';
}
if (!(CInputCounter as Story).storyName) {
  (CInputCounter as Story).storyName = 'Input counter';
}
if (!(CAutofocusScreenshot as Story).storyName) {
  (CAutofocusScreenshot as Story).storyName = 'Autofocus screenshot';
}
if (!(CDisabled as Story).storyName) {
  (CDisabled as Story).storyName = 'Disabled';
}
if (!(CExtraSmallContainer as Story).storyName) {
  (CExtraSmallContainer as Story).storyName = 'Extra small container';
}
if (!(CSizeScreenshot as Story).storyName) {
  (CSizeScreenshot as Story).storyName = 'Size screenshot';
}
if (!(CSuffixAndPrefixScreenshot as Story).storyName) {
  (CSuffixAndPrefixScreenshot as Story).storyName = 'Suffix and prefix screenshot';
}
if (!(CTextareaDisabled as Story).storyName) {
  (CTextareaDisabled as Story).storyName = 'Textarea disabled';
}
if (!(CTextareaHeight as Story).storyName) {
  (CTextareaHeight as Story).storyName = 'Textarea height';
}

export default {
  title: 'Components/Input/截图测试',
  tags: ['autodocs'],
};

export {
  C999Case0,
  CInputCounter,
  CAutofocusScreenshot,
  CDisabled,
  CExtraSmallContainer,
  CSizeScreenshot,
  CSuffixAndPrefixScreenshot,
  CTextareaDisabled,
  CTextareaHeight,
};
