// @universe-design/react/src/input/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import CAddon from '../demo/addon';
import CBasic from '../demo/basic';
import CCounter from '../demo/counter';
import CFormat from '../demo/format';
import CGroup from '../demo/group';
import CPassword from '../demo/password';
import CPrefixSuffix from '../demo/prefix-suffix';
import CSearch from '../demo/search';
import CSize from '../demo/size';
import CStatus from '../demo/status';
import CTextareaResize from '../demo/textarea-resize';
import CTextareaSuffix from '../demo/textarea-suffix';
import CTextarea from '../demo/textarea';
if (!(CAddon as Story).storyName) {
  (CAddon as Story).storyName = '2. 前置 / 后置标签';
  (CAddon as Story).parameters = {
    docs: {
      description: {
        story: '通过设置`addonBefore`和`addonAfter`，在输入框前后添加元素。\n\n',
      },
    },
  };
}
if (!(CBasic as Story).storyName) {
  (CBasic as Story).storyName = '0. 基础用法';
  (CBasic as Story).parameters = {
    docs: {
      description: {
        story: '简单的用法。\n\n',
      },
    },
  };
}
if (!(CCounter as Story).storyName) {
  (CCounter as Story).storyName = '9. 字数统计';
  (CCounter as Story).parameters = {
    docs: {
      description: {
        story:
          '`Input` 和 `Input.TextArea` 组件都采取以下字数统计和限制输入规则：\n当仅设置 maxLength 为数字时，会添加原生属性 `maxLength`, 此时使用浏览器原生能力来限制输入长度。\n设置 `showCount` 可以开启字数统计，默认会按照使用 unicode 码元个数来计数 ([UD 的计数规则](https://bytedance.feishu.cn/docx/doxcnox8gbqOZ5zZ1U2kg27inOT))。也可以使用 `counter` 来自定义计数逻辑。\n当设置 `showCount` 和 `maxLength` 时，不会截断超出字数部分，如果需要截断，可设置`truncateExcessWords`属性。\n\n** counter 和 slicer **\n\n当需要自定义计数规则时，需要告诉组件如何去计数，以及计数超过 maxLength 之后，如何做截断。此时需要使用 `counter` 和 `slicer`。\n',
      },
    },
  };
}
if (!(CFormat as Story).storyName) {
  (CFormat as Story).storyName = '4. 输入时格式化展示';
  (CFormat as Story).parameters = {
    docs: {
      description: {
        story: '结合 Tooltip 组件，实现一个数值输入框，方便内容超长时的全量展现。\n\n',
      },
    },
  };
}
if (!(CGroup as Story).storyName) {
  (CGroup as Story).storyName = '6. 输入框组合';
  (CGroup as Story).parameters = {
    docs: {
      description: {
        story: '输入框的组合展现。\n\n',
      },
    },
  };
}
if (!(CPassword as Story).storyName) {
  (CPassword as Story).storyName = '5. 密码输入';
  (CPassword as Story).parameters = {
    docs: {
      description: {
        story: '用于密码的输入。\n\n',
      },
    },
  };
}
if (!(CPrefixSuffix as Story).storyName) {
  (CPrefixSuffix as Story).storyName = '3. 前缀 / 后缀';
  (CPrefixSuffix as Story).parameters = {
    docs: {
      description: {
        story: '通过设置`prefix`和`suffix`为输入框前后添加元素。\n\n',
      },
    },
  };
}
if (!(CSearch as Story).storyName) {
  (CSearch as Story).storyName = '7. 搜索框';
  (CSearch as Story).parameters = {
    docs: {
      description: {
        story: '带有搜索按钮的输入框，用于内容检索。\n\n',
      },
    },
  };
}
if (!(CSize as Story).storyName) {
  (CSize as Story).storyName = '1. 三种尺寸';
  (CSize as Story).parameters = {
    docs: {
      description: {
        story: '输入框有 3 种尺寸，分别为`large`、`medium`、`small`，默认尺寸为`medium`。\n\n',
      },
    },
  };
}
if (!(CStatus as Story).storyName) {
  (CStatus as Story).storyName = '0. 状态';
  (CStatus as Story).parameters = {
    docs: {
      description: {
        story: '错误和禁用状态。\n\n',
      },
    },
  };
}
if (!(CTextareaResize as Story).storyName) {
  (CTextareaResize as Story).storyName = '10. 文本域 - 适应文本高度';
  (CTextareaResize as Story).parameters = {
    docs: {
      description: {
        story: '设置 `autoSize`，文本域会自动根据输入的文本调整文本域的高度。\n\n',
      },
    },
  };
}
if (!(CTextareaSuffix as Story).storyName) {
  (CTextareaSuffix as Story).storyName = '11. 文本域 - 后缀';
  (CTextareaSuffix as Story).parameters = {
    docs: {
      description: {
        story: '设置 `suffix`，可以给文本域设置后缀，比如设置语言标签。\n\n',
      },
    },
  };
}
if (!(CTextarea as Story).storyName) {
  (CTextarea as Story).storyName = '8. 文本域';
  (CTextarea as Story).parameters = {
    docs: {
      description: {
        story: '用于多行输入场景。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Input/官网示例',
  tags: ['autodocs'],
};

export {
  CAddon,
  CBasic,
  CCounter,
  CFormat,
  CGroup,
  CPassword,
  CPrefixSuffix,
  CSearch,
  CSize,
  CStatus,
  CTextareaResize,
  CTextareaSuffix,
  CTextarea,
};
