// @universe-design/react/src/progress/Line.tsx
import type { FC, CSSProperties, ReactNode } from 'react';
import React from 'react';
import { useCssScope } from '../config-provider/hooks';
import clsx from 'clsx';
import Steps from './Steps';
import type { ProgressProps, ProgressStatusType } from './shared';

interface LineProps extends ProgressProps {
  percent: number;
  strokeColor?: string;
  /**
   * 进度条线宽度，单位 px
   * @defaultValue 4
   */
  strokeWidth?: number;
  format?: (percent: number) => ReactNode;
}

const Line: FC<LineProps> = (props, refs) => {
  const {
    percent,
    strokeColor = '',
    strokeWidth = 4,
    status,
    trailColor,
    showInfo = true,
    infoPosition = 'right',
    format,
  } = props;

  let lineStatus: ProgressStatusType = status || 'normal';

  if (lineStatus === 'normal' && percent >= 100) {
    lineStatus = 'success';
  }

  const cssScope = useCssScope('progress');

  const renderSteps = () => (
    <span className={cssScope('--steps')}>
      <Steps percent={percent} status={lineStatus} format={format} />
    </span>
  );

  const clsBg = clsx([cssScope(), cssScope('--line-bg'), cssScope(`--line-bg-${lineStatus}`)]);

  const bgStyle = {
    width: `${percent}%`,
    [strokeColor && 'backgroundColor']: strokeColor,
  } as CSSProperties;

  const clsInnerLine = clsx([
    cssScope(),
    cssScope('--line-inner'),
    { [cssScope(`--line-inner-with-${infoPosition}-info`)]: showInfo },
  ]);

  const innerStyle = {
    height: strokeWidth,
  } as CSSProperties;

  if (trailColor) {
    innerStyle.backgroundColor = trailColor;
  }

  return (
    <>
      {showInfo && infoPosition === 'left' && renderSteps()}
      <div className={clsInnerLine} style={innerStyle}>
        <div className={clsBg} style={bgStyle} />
      </div>
      {showInfo && infoPosition === 'right' && renderSteps()}
    </>
  );
};

export default Line;
