// @universe-design/react/src/progress/shared.ts
import type { CSSProperties, ReactNode } from 'react';
import type { ElementOf } from '../util';

export const ProgressTypesEnum = ['circle', 'line'] as const;
export const InfoPositionTypesEnum = ['left', 'right'] as const;
export const ProgressStatusEnum = ['success', 'exception', 'normal'] as const;

export type ProgressType = ElementOf<typeof ProgressTypesEnum>;
export type InfoPositionType = ElementOf<typeof InfoPositionTypesEnum>;
export type ProgressStatusType = ElementOf<typeof ProgressStatusEnum>;

/**
 * @public
 */
export interface ProgressProps {
  /**
   * 进度条百分比
   * @defaultValue 0
   */
  percent?: number;
  /**
   * 是否显示进度数值或状态图标
   * @defaultValue true
   */
  showInfo?: boolean;
  /**
   * 进度条的色彩
   */
  strokeColor?: string;
  /**
   * 进度条类型
   * @defaultValue `line`
   */
  type?: ProgressType;
  /**
   * 进度值位置
   * @defaultValue "right"
   */
  infoPosition?: InfoPositionType;
  /**
   * 进度条线的宽度，单位为`px`
   * @defaultValue 4
   */
  strokeWidth?: number;
  /**
   * 进度条状态
   */
  status?: ProgressStatusType;
  /**
   * 未完成的分段的颜色
   */
  trailColor?: string;
  /**
   * 圆形进度条线的宽度，单位为`px`
   * @defaultValue 16
   */
  width?: number;
  /**
   * 透传属性
   */
  className?: string;
  /**
   * 透传属性
   */
  style?: CSSProperties;
  /**
   * 自定义进度描述
   */
  format?: (percent: number) => ReactNode;
}
