// @universe-design/react/src/notice/index.tsx
import ErrorFilled from '@universe-design/icons-react/ErrorFilled';
import CloseOutlined from '@universe-design/icons-react/CloseOutlined';
import InfoFilled from '@universe-design/icons-react/InfoFilled';
import WarningFilled from '@universe-design/icons-react/WarningFilled';
import SucceedFilled from '@universe-design/icons-react/SucceedFilled';
import stylex from '@universe-design/stylex';
import { useControlledState } from '@universe-design/react-utilities';
import { Button } from '../button';
import type { ButtonProps } from '../button';
import clsx from 'clsx';
import type { HTMLAttributes } from 'react';
import React, { useCallback, useRef } from 'react';
import type { PresetStatusColorType } from '../util';
import type { Slot } from '../shared/slot';
import { resolveSlot, renderSlotNode } from '../shared/slot';
import { devWarning, warning } from '../util';
import { useCssScope } from '../config-provider/hooks';
import { SpinLoadingSvg } from '../_assets/svgs';
import { cssTypographyHeadline } from '../stylex';

import './style';

const iconMap = {
  success: SucceedFilled,
  warning: WarningFilled,
  error: ErrorFilled,
  loading: SpinLoadingSvg,
  info: InfoFilled,
};

export interface NoticeProps {
  /**
   * 设置主题，outlined 包含边框，filled 为无边框
   * @internal
   */
  theme?: 'outlined' | 'filled' | 'plain';
  /**
   * 状态类型
   * @defaultValue `info`
   */
  type?: PresetStatusColorType | 'loading';
  /**
   * 是否可关闭
   * @defaultValue false
   */
  closable?: boolean;
  /**
   * 关闭按钮，UD 不建议使用
   * @deprecated
   */
  closeText?: React.ReactNode;
  /**
   * 关闭按钮
   */
  closeButton?: Slot<ButtonProps>;
  /**
   * 提示内容，如果有`description`则会变为粗体样式
   */
  message: React.ReactNode;
  /**
   * 提示辅助描述
   */
  description?: React.ReactNode;
  /**
   * 跟在提示内容后面的操作按钮
   */
  action?: React.ReactNode;
  /**
   * 可见状态变化回调
   */
  onClose?: React.MouseEventHandler<HTMLSpanElement | HTMLButtonElement>;
  /**
   * UD 不支持 afterClose
   * @deprecated
   */
  afterClose?: () => void;
  /**
   * 是否显示图标，这里和其他开源框架不同，UD 默认为 true。
   * @defaultValue true
   */
  showIcon?: boolean;
  /**
   * [aria-role-attribute](https://www.w3.org/TR/2014/REC-html5-20141028/dom.html#aria-role-attribute)。
   * @defaultValue `alert`
   */
  role?: string;
  /**
   *  替换默认的 class 前缀
   */
  prefixCls?: string;
  /**
   * 是否为无边框，无 margin 模式
   * @defaultValue false
   */
  banner?: boolean;
  /**
   * 替换状态按钮。
   */
  icon?: React.ReactNode;
  /**
   * 可见状态的受控值。
   */
  visible?: boolean;
  /**
   * 标题对齐方式。在同时有`message`和`description`属性时不生效。
   * @defaultValue `left`
   */
  align?: 'left' | 'center';
  /**
   * 最大行数，默认 4
   */
  maxLine?: number;
}

export const Notice = React.forwardRef<HTMLDivElement, HTMLAttributes<HTMLDivElement> & NoticeProps>(function (
  props,
  ref,
) {
  const {
    className,
    closable = false,
    align = 'left',
    type = 'info',
    theme = 'filled',
    message,
    description,
    action,
    visible: visibleProp,
    banner = false,
    onClose,
    closeText,
    closeButton,
    afterClose,
    showIcon = true,
    role = 'alert',
    icon,
    prefixCls,
    maxLine,
    ...restProps
  } = props;

  warning(
    !afterClose,
    devWarning(Notice.displayName || Notice.name, 'Close motion is not support, so afterClose is not provided.'),
  );

  const cssScope = useCssScope('notice', prefixCls);
  const [visible, setVisible] = useControlledState(true, visibleProp);
  const descRef = useRef<HTMLDivElement>(null);
  const isNoTitle = message && !description;

  let iconNode = null;
  if (showIcon) {
    if (icon) {
      iconNode = <div className={cssScope('__statusIcon')}>{icon}</div>;
    }
    if (!iconNode) {
      const StatusIcon = iconMap[type] || iconMap.info;
      iconNode = <StatusIcon className={cssScope('__statusIcon')} />;
    }
  }
  const isClosable = closeText ? true : closable;

  const handleClick: React.MouseEventHandler<HTMLSpanElement | HTMLButtonElement> = useCallback(
    e => {
      setVisible(false);
      onClose && onClose(e);
    },
    [setVisible, onClose],
  );

  const closeBtn = React.useMemo(() => {
    if (closeText) {
      return (
        <span className={cssScope('__close')} onClick={handleClick}>
          {closeText}
        </span>
      );
    }

    const closeButtonSlot = resolveSlot(closeButton, {
      elementType: Button,
      defaultProps: { icon: <CloseOutlined />, size: 'small', type: 'icon' },
      renderByDefault: true,
    });

    if (closeButtonSlot) {
      closeButtonSlot.props.className = clsx(cssScope('__close'), closeButtonSlot.props.className);
      const originalOnClick = closeButtonSlot.props.onClick;
      closeButtonSlot.props.onClick = (e: React.MouseEvent<HTMLElement>) => {
        originalOnClick?.(e);
        handleClick(e);
      };
    }

    return renderSlotNode(closeButtonSlot);
  }, [closeButton, closeText, cssScope, handleClick]);

  const innerType = type === 'loading' ? 'info' : type;

  const contentStyle = maxLine ? { WebkitLineClamp: maxLine } : undefined;

  return visible ? (
    <div
      ref={ref}
      className={clsx(
        cssScope(),
        cssScope(`-${innerType}`),
        align === 'center' && isNoTitle && cssScope('-align-center'),
        closable && cssScope('-closable'),
        banner && cssScope('-banner'),
        iconNode && cssScope('-showIcon'),
        action && cssScope('-has-action'),
        message && description && cssScope('-has-title'),
        cssScope(`--${theme}`),
        className,
      )}
      role={role}
      {...restProps}
    >
      {iconNode}
      <div className={clsx(cssScope('__main'), isNoTitle && cssScope('__main--no-title'))}>
        {message && description && (
          <div className={cssScope('__title')}>
            <div className={clsx(cssScope('__title__content'), stylex(cssTypographyHeadline))}>{message}</div>
            {isClosable && closeBtn}
          </div>
        )}
        {(description || isNoTitle) && (
          <div ref={descRef} className={cssScope('__description')}>
            <div className={cssScope('__description-content')} style={contentStyle}>
              {description || message}
            </div>
            {action && <div className={cssScope('__description-action')}>{action}</div>}
          </div>
        )}
        {isClosable && isNoTitle && closeBtn}
      </div>
    </div>
  ) : null;
});
