// @universe-design/react/lib/overflow/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Overflow = void 0;
var _clsx = _interopRequireDefault(require("clsx"));
var _react = _interopRequireDefault(require("react"));
var _hooks = require("../config-provider/hooks");
var _hooks2 = require("./hooks");
require("./style");
const _OverflowItem = _ref => {
  let {
    value,
    index,
    children,
    getItemStyle
  } = _ref;
  const cssScope = (0, _hooks.useCssScope)('overflow__item');
  const {
    ref,
    isVisible
  } = (0, _hooks2.useOverflowItem)({
    key: value,
    index
  });
  const itemStyle = getItemStyle === null || getItemStyle === void 0 ? void 0 : getItemStyle(value);
  const order = isVisible ? index : Number.MAX_SAFE_INTEGER;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: cssScope(),
    style: {
      order,
      height: isVisible ? undefined : 0,
      overflowY: isVisible ? undefined : 'hidden',
      pointerEvents: isVisible ? undefined : 'none',
      ...itemStyle
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    ref: ref
  }, children));
};
if (process.env.NODE_ENV !== "production") {
  _OverflowItem.displayName = 'OverflowItem';
}
const OverflowItem = /*#__PURE__*/_react.default.memo(_OverflowItem);
const OverflowRest = /*#__PURE__*/_react.default.memo(_ref2 => {
  let {
    children,
    restStyle
  } = _ref2;
  const {
    ref,
    countBeforeRest,
    isVisible,
    isResponsive,
    restPosition
  } = (0, _hooks2.useOverflowRest)();
  const cssScope = (0, _hooks.useCssScope)('overflow__rest');
  const order = isVisible ? countBeforeRest - 1 : Number.MAX_SAFE_INTEGER;
  return /*#__PURE__*/_react.default.createElement("div", {
    ref: ref,
    className: cssScope(''),
    style: {
      // 这里是为了修复在父级 fit-content 时，rest 节点没有展示也会占空间的问题
      position: isVisible ? undefined : 'absolute',
      order: isResponsive ? order : restPosition - 1,
      height: isVisible ? undefined : 0,
      overflowY: isVisible ? undefined : 'hidden',
      pointerEvents: isVisible ? undefined : 'none',
      ...restStyle
    }
  }, children);
});
const OverflowSuffix = /*#__PURE__*/_react.default.memo(_ref3 => {
  let {
    className,
    children
  } = _ref3;
  const {
    ref,
    suffixFixedStart,
    countAfterRest,
    countBeforeRest,
    isResponsive,
    restWidth,
    suffixWrap
  } = (0, _hooks2.useOverflowSuffix)();
  const cssScope = (0, _hooks.useCssScope)('overflow__suffix');
  return /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(className, cssScope('')),
    ref: ref,
    style: {
      order: isResponsive ? countBeforeRest + countAfterRest - 1 : Number.MAX_SAFE_INTEGER,
      ...(suffixFixedStart !== null && isResponsive ? {
        position: 'absolute',
        left: suffixFixedStart
      } : {}),
      maxWidth: suffixWrap ? undefined : `calc(100% - ${restWidth}px)` // 限制 suffix 不换行时的最大宽度
    }
  }, children);
});
const defaultItemRenderer = key => /*#__PURE__*/_react.default.createElement("div", {
  style: {
    border: '1px red solid'
  }
}, Array(Number(key) + 1).fill(0).map((_, i) => i).join(','));
const defaultRestRenderer = omittedKeys => /*#__PURE__*/_react.default.createElement("div", {
  style: {
    border: '1px blue solid'
  }
}, `+${omittedKeys.length}`);
const Overflow = _ref4 => {
  let {
    maxCount,
    keys,
    style,
    getItemStyle,
    restStyle,
    className,
    itemRenderer = defaultItemRenderer,
    restRenderer = defaultRestRenderer,
    suffixRenderer,
    suffixClassName,
    restPosition,
    addonRenderer,
    alwaysObserveResize = false,
    suffixWrap = true
  } = _ref4;
  const cssScope = (0, _hooks.useCssScope)('overflow');
  const {
    ref,
    dispatch,
    isResponsive,
    ...state
  } = (0, _hooks2.useOverflow)({
    maxCount,
    keys,
    hasSuffix: Boolean(suffixRenderer),
    restPosition,
    suffixWrap
  });
  const showRest = isResponsive || typeof maxCount === 'number' && keys.length > maxCount;
  return /*#__PURE__*/_react.default.createElement(_hooks2.OverflowContext.Provider, {
    value: {
      dispatch,
      isResponsive,
      ...state,
      alwaysObserveResize
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    ref: ref,
    className: (0, _clsx.default)(cssScope(), className),
    style: style
  }, [...state.mergedKeys.beforeRest, ...state.mergedKeys.afterRest].map((key, i) => /*#__PURE__*/_react.default.createElement(OverflowItem, {
    key: key,
    value: key,
    index: i,
    getItemStyle: getItemStyle
  }, itemRenderer(key))), showRest && /*#__PURE__*/_react.default.createElement(OverflowRest, {
    restStyle: restStyle
  }, restRenderer(state.omittedKeys)), suffixRenderer && /*#__PURE__*/_react.default.createElement(OverflowSuffix, {
    className: suffixClassName
  }, suffixRenderer({
    omittedKeys: state.omittedKeys
  })), addonRenderer === null || addonRenderer === void 0 ? void 0 : addonRenderer({
    omittedKeys: state.omittedKeys
  })));
};
exports.Overflow = Overflow;
