// @universe-design/react/lib/carousel/transition-descriptors/fade.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FadeCarouselTransitionDescriptor = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _timingFunctionConsts = require("../../shared/components/transition/timingFunctionConsts");
var _LinkedList = require("../../util/datastructure/LinkedList");
var _utils = require("./utils");
class FadeCarouselTransitionDescriptor {
  constructor() {
    var _params$transitionTim;
    let params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    (0, _defineProperty2.default)(this, "defaults", {
      duration: 1000
    });
    (0, _defineProperty2.default)(this, "_transitionTimingFunction", void 0);
    this._transitionTimingFunction = (_params$transitionTim = params.transitionTimingFunction) !== null && _params$transitionTim !== void 0 ? _params$transitionTim : _timingFunctionConsts.transitionTimingFunctionStandard;
  }
  onGotoSlicklyStart(params, manager) {
    const {
      slide,
      element
    } = params;
    const {
      state: {
        slide: curSlide,
        prevSlide
      }
    } = manager;
    if (prevSlide && (0, _utils.isSlideEqual)(slide, prevSlide) || !prevSlide && (0, _utils.isSlideEqual)(slide, curSlide)) {
      this._show(element);
    } else {
      this._hide(element);
    }
    this._setPosition(element, this._getSlidePosition(slide, prevSlide !== null && prevSlide !== void 0 ? prevSlide : curSlide, manager));
  }
  onGotoSlicklyActive(params, manager) {
    const {
      slide,
      element
    } = params;
    const {
      duration,
      state: {
        slide: curSlide
      }
    } = manager;
    element.style.willChange = 'opacity';
    element.style.transition = `opacity ${duration}ms ${this._transitionTimingFunction}`;
    if ((0, _utils.isSlideEqual)(slide, curSlide)) {
      this._show(element);
    } else {
      this._hide(element);
    }
  }
  onGotoSlicklyDone(params, manager) {
    const {
      slide,
      element
    } = params;
    const {
      state: {
        slide: curSlide
      }
    } = manager;
    element.style.willChange = '';
    element.style.transition = '';
    if ((0, _utils.isSlideEqual)(slide, curSlide)) {
      this._show(element);
    } else {
      this._hide(element);
    }
  }
  onGotoDirectly(params, manager) {
    const {
      slide,
      element
    } = params;
    const {
      state: {
        slide: curSlide
      }
    } = manager;
    if ((0, _utils.isSlideEqual)(slide, curSlide)) {
      this._show(element);
    } else {
      this._hide(element);
    }
    this._setPosition(element, this._getSlidePosition(slide, curSlide, manager));
  }
  getSlidesTOBeRendered(manager) {
    // Constant
    const cacheCount = 1;
    // State
    const {
      slideCount,
      state: {
        slide: curSlide
      }
    } = manager;
    const slides = new _LinkedList.LinkedList();
    for (let i = -cacheCount; i <= cacheCount; i++) {
      const slide = (0, _utils.getSlideFromAbsoluteIndex)((0, _utils.getAbsoluteIndexFromSlide)(curSlide, slideCount) + i, slideCount);
      slides.append(slide);
    }
    return slides.toArray();
  }
  _getSlidePosition(slide, curSlide, manager) {
    const {
      state: {
        windowSize
      }
    } = manager;
    return {
      x: windowSize.width / 2 - manager.getSlideSize(slide.slideIndex).width / 2,
      y: 0,
      z: 0
    };
  }
  _setPosition(element, position) {
    const {
      x,
      y,
      z
    } = position;
    element.style.transform = `translate3d(${x}px, ${y}px, ${z}px)`;
  }
  _show(element) {
    element.style.opacity = '1';
    element.style.pointerEvents = '';
  }
  _hide(element) {
    element.style.opacity = '0';
    element.style.pointerEvents = 'none';
  }
}
exports.FadeCarouselTransitionDescriptor = FadeCarouselTransitionDescriptor;
