// @universe-design/react/lib/carousel/transition-descriptors/push.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PushCarouselTransitionDescriptor = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _timingFunctionConsts = require("../../shared/components/transition/timingFunctionConsts");
var _LinkedList = require("../../util/datastructure/LinkedList");
var _utils = require("./utils");
class PushCarouselTransitionDescriptor {
  constructor() {
    var _params$transitionTim;
    let params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    (0, _defineProperty2.default)(this, "defaults", {
      duration: 300
    });
    (0, _defineProperty2.default)(this, "_transitionTimingFunction", void 0);
    (0, _defineProperty2.default)(this, "_setPosition", (element, position) => {
      const {
        x,
        y,
        z
      } = position;
      element.style.transform = `translate3d(${x}px, ${y}px, ${z}px)`;
    });
    (0, _defineProperty2.default)(this, "_getSlidePosition", (slide, curSlide, manager) => {
      const {
        slideCount,
        state: {
          windowSize
        }
      } = manager;
      const slideAbsoluteIndex = (0, _utils.getAbsoluteIndexFromSlide)(slide, slideCount);
      const curSlideAbsoluteIndex = (0, _utils.getAbsoluteIndexFromSlide)(curSlide, slideCount);
      // -7 [  4  ] -3 [  6  ] 3 [  2  ]
      // pos(n) - pos(n - 1) = w(n - 1)
      // pos(0) = - w(0) / 2
      const position = {
        x: 0,
        y: 0,
        z: 0
      };
      if (slideAbsoluteIndex > curSlideAbsoluteIndex) {
        for (let i = curSlideAbsoluteIndex + 1; i <= slideAbsoluteIndex; i++) {
          const {
            slideIndex
          } = (0, _utils.getSlideFromAbsoluteIndex)(i - 1, slideCount);
          position.x += manager.getSlideSize(slideIndex).width;
        }
      } else {
        for (let i = curSlideAbsoluteIndex - 1; i >= slideAbsoluteIndex; i--) {
          const {
            slideIndex
          } = (0, _utils.getSlideFromAbsoluteIndex)(i, slideCount);
          position.x -= manager.getSlideSize(slideIndex).width;
        }
      }
      position.x -= manager.getSlideSize(curSlide.slideIndex).width / 2; // 减去中心 slide 的一半宽度
      position.x += windowSize.width / 2; // 加上一半的视窗宽度
      return position;
    });
    this._transitionTimingFunction = (_params$transitionTim = params.transitionTimingFunction) !== null && _params$transitionTim !== void 0 ? _params$transitionTim : _timingFunctionConsts.transitionTimingFunctionStandard;
  }
  getSlidesTOBeRendered(manager) {
    // Constant

    // State
    const {
      slideCount,
      state: {
        slide: curSlide,
        windowSize
      }
    } = manager;
    const slides = new _LinkedList.LinkedList([curSlide]);
    const curSlideWidth = manager.getSlideSize(curSlide.slideIndex).width;
    // 从中线开始往两边扩散，停止扩散的条件
    // 1. 遇到当前 slide 宽度是 0 说明还没有渲染过，停止扩散
    // 2. 大于等于 cacheWindowTimes / 2 倍的 windowSize 时，停止扩散
    for (const direction of [-1, 1]) {
      for (let _width = curSlideWidth / 2, _slide = curSlide;;) {
        if (_width >= 3 / 2 * windowSize.width) {
          break;
        }
        _slide = (0, _utils.getSlideFromAbsoluteIndex)((0, _utils.getAbsoluteIndexFromSlide)(_slide, slideCount) + direction, slideCount);
        const _slideSize = manager.getSlideSize(_slide.slideIndex);
        if (direction > 0) {
          slides.append(_slide);
        } else {
          slides.prepend(_slide);
        }
        if (_slideSize.width === 0) {
          break;
        }
        _width += _slideSize.width;
      }
    }
    return slides.toArray();
  }
  onGotoSlicklyStart(params, manager) {
    const {
      slide,
      element
    } = params;
    const {
      state: {
        slide: curSlide,
        prevSlide
      }
    } = manager;
    this._setPosition(element, this._getSlidePosition(slide, prevSlide !== null && prevSlide !== void 0 ? prevSlide : curSlide, manager));
  }
  onGotoSlicklyActive(params, manager) {
    const {
      slide,
      element
    } = params;
    const {
      duration,
      state: {
        slide: curSlide
      }
    } = manager;
    element.style.willChange = 'transform';
    element.style.transition = `transform ${duration}ms ${this._transitionTimingFunction}`;
    this._setPosition(element, this._getSlidePosition(slide, curSlide, manager));
  }
  onGotoSlicklyDone(params) {
    const {
      element
    } = params;
    element.style.willChange = '';
    element.style.transition = '';
  }
  onGotoDirectly(params, manager) {
    const {
      slide,
      element
    } = params;
    const {
      state: {
        slide: curSlide
      }
    } = manager;
    this._setPosition(element, this._getSlidePosition(slide, curSlide, manager));
  }
}
exports.PushCarouselTransitionDescriptor = PushCarouselTransitionDescriptor;
