// @universe-design/react/src/date-picker/Context.ts
import type moment from 'moment';
import type { Dispatch, ReactNode, SetStateAction } from 'react';
import { createContext, useContext, useContextSelector } from 'use-context-selector';
import type { SharedTimeProps } from '../shared/components/picker/interface';
import { devWarning, warning } from '../util';
import type { MODE } from './interface';

export interface DatePickerContextType {
  /**
   * 日期面板的状态
   */
  mode: MODE;
  /**
   * 设置日期面板的状态
   */
  setMode: Dispatch<SetStateAction<MODE>>;
  /**
   * 选中的时间
   */
  value?: moment.Moment;
  /**
   * 设置选中的时间
   */
  setValue: (value: SetStateAction<moment.Moment | undefined>, autoCommit?: boolean) => void;
  /**
   * 设置弹窗开启状态
   */
  setOpened: Dispatch<SetStateAction<boolean>>;
  /**
   * 是否展示今天按钮
   */
  showToday: boolean;
  /**
   * 是否展示当前时间按钮
   * @defaultValue false
   */
  showNow?: boolean;
  /**
   * @defaultValue false
   * 是否显示时间选择
   */
  showTime?: boolean | SharedTimeProps;
  /**
   * 是否需要二次确认
   * @defaultValue false
   */
  needConfirm?: boolean;
  /**
   * 自定义 footer
   */
  renderExtraFooter?: (ctx: Pick<DatePickerContextType, 'mode' | 'value' | 'setValue' | 'setOpened'>) => ReactNode;
}

const DatePickerContext = createContext<DatePickerContextType>({
  mode: 'date',
  setMode: () => warning(false, devWarning('DatePicker', 'You are using the default setMode from context')),
  setValue: () => warning(false, devWarning('DatePicker', 'You are using the default setValue from context')),
  setOpened: () => warning(false, devWarning('DatePicker', 'You are using the default setOpened from context')),
  showToday: true,
  showTime: false,
  showNow: false,
  needConfirm: false,
});

export const useDatePickerContext = () => useContext(DatePickerContext);

export const useDatePickerContextSelector = <Selected>(
  selector: (value: DatePickerContextType) => Selected,
): Selected => useContextSelector(DatePickerContext, selector);

export const { Provider: DatePickerContextProvider } = DatePickerContext;
