// @universe-design/react/src/date-picker/DatePicker.tsx
import CalendarOutlined from '@universe-design/icons-react/CalendarOutlined';
import type { TriggerProps } from '@universe-design/rc-trigger';
import type { SharedTimeProps } from '../shared/components/picker';
import clsx from 'clsx';
import noop from 'lodash/noop';
import moment from 'moment';
import type { CSSProperties, ReactNode, SetStateAction } from 'react';
import React, { useCallback, useMemo, useRef, useState } from 'react';
import { useControlledState } from '@universe-design/react-utilities';
import {
  useComponentSize,
  useCssScope,
  useDirection,
  useInputGroupSize,
  useLocaleReceiver,
} from '../config-provider/hooks';
import useControlled from '../hooks/useControlled';
import useSyncState from '../hooks/useSyncState';
import NativeInput from '../shared/components/Input';
import { ListBox } from '../shared/components/list/List';
import type { DefaultSizeType } from '../util';
import { devWarning, getDefaultSizeAbbr, isFunction, warning } from '../util';
import composeRefs from '../util/composeRefs';
import type { DatePickerContextType } from './Context';
import { DatePickerContextProvider } from './Context';
import type { IFormat, IParse, MODE, PickerType, StartOfWeek } from './interface';
import PickerPanel, { isDateValueEqual } from './panels/picker';
import { isEqual } from './utils/date';
import { getInputSize } from './utils/ui';
import generateConfig from '../shared/components/picker/generate/moment';

import './style';
import { usePopperSidecar } from '../config-provider/hooks/usePopperSidecar';
import type { PopperShowActionType } from '../shared';
import { SlideUpDownTransition } from '../shared/components/transition/SlideUpDownTransition';
import useClickOutsideEffect from '../shared/components/picker/hooks/useClickOutsideEffect';
import InputWrapper from '../shared/components/picker/Comp/input-wrapper';
import useFormatAndParser from './hooks/useFormatAndParser';
import type { DateCalendarUIProps } from './panels/DateCalendar/interface';
import { isInputDateDisabled } from './utils/disabled';
import { usePopupPortalContainer } from '@universe-design/react-provider';

export interface DatePickerState {
  /**
   * 选中后计算出的文本
   **/
  text: string;
  /**
   * 用户 hover 后计算出的文本
   */
  hoverText: string;
  /**
   * 选中的值
   */
  value: moment.Moment | undefined;
  /**
   * 是否展开弹层
   */
  open: boolean;
}

export type DatePickerTriggerElement = ReactNode | ((state: DatePickerState) => ReactNode);

export interface DatePickerProps
  extends Pick<
    DateCalendarUIProps,
    'minDate' | 'maxDate' | 'disabledDate' | 'mode' | 'defaultPickerValue' | 'hideWeekNumbers' | 'janX'
  > {
  /**
   * @defaultValue `date`
   * @deprecated 请使用 mode
   */
  picker?: PickerType;
  /**
   * `className` 属性透传
   */
  className?: string;
  /**
   * `style` 属性透传
   */
  style?: CSSProperties;
  /**
   * `id`属性透传
   */
  id?: string;
  /**
   * 自定义的选择框后缀图标
   */
  suffixIcon?: React.ReactNode;
  /**
   * 是否显示清除按钮
   * @defaultValue true
   */
  allowClear?: boolean;
  /**
   * 自动获取焦点
   * @defaultValue false
   */
  autoFocus?: boolean;
  /**
   * 禁用选日期范围
   * @defaultValue false
   */
  disabled?: boolean;
  /**
   * 设置输入框为只读（避免在移动设备上打开虚拟键盘）
   * @defaultValue false
   */
  inputReadOnly?: boolean;
  /**
   * 输入框提示文字
   */
  placeholder?: string;
  /**
   * 弹层元素的 `style` 属性透传
   */
  popupStyle?: CSSProperties;
  /**
   * 选中的日期的受控值
   */
  value?: moment.Moment;
  /**
   * 选中的日期的默认值
   */
  defaultValue?: moment.Moment;
  /**
   * 选中日期值变化的回调
   */
  onChange?: (value: moment.Moment | undefined) => void;
  /**
   * 日期显示格式，见 [Moment](https://momentjs.com/docs/#/displaying/)
   */
  format?: IFormat;
  /**
   * 自定义解析 input 字符串的方法
   */
  parse?: IParse;
  /**
   * 控制弹层是否展开的受控值
   */
  open?: boolean;
  /**
   * 控制弹层是否展开的默认值
   * @defaultValue false
   */
  defaultOpen?: boolean;
  /**
   * 输入框大小
   * @defaultValue `medium`
   */
  size?: DefaultSizeType;
  /**
   * 是否是错误状态
   * @defaultValue false
   */
  error?: boolean;
  /**
   * 替换触发元素
   */
  triggerElement?: DatePickerTriggerElement;
  /**
   * 定义浮层的容器，默认为 body 上新建 div
   */
  getPopupContainer?: TriggerProps['getPopupContainer'];
  /**
   * 弹出日历和关闭日历的回调
   */
  onOpenChange?: (open: boolean) => void;
  /**
   * 原生事件
   */
  onKeyDown?: React.KeyboardEventHandler<HTMLInputElement>;
  /**
   * 原生事件
   */
  onMouseDown?: React.MouseEventHandler<HTMLDivElement>;
  /**
   * 原生事件
   */
  onMouseUp?: React.MouseEventHandler<HTMLDivElement>;
  /**
   * 原生事件
   */
  onClick?: React.MouseEventHandler<HTMLDivElement>;
  /**
   * 点击选中日期的回调
   */
  onSelect?: (value?: moment.Moment) => void;
  /**
   * 在面板中添加额外的页脚
   */
  renderExtraFooter?: DatePickerContextType['renderExtraFooter'];
  /**
   * 是否展示“今天”按钮
   * @defaultValue true
   */
  showToday?: boolean;
  /**
   * 弹层元素的 `className` 属性透传
   */
  dropdownClassName?: string;
  /**
   * 自定义日期单元格的内容
   */
  dateRender?: (currentDate: moment.Moment, mode?: MODE, defaultNode?: React.ReactNode) => React.ReactNode;
  /**
   * 日期面板变化时的回调，用于配合 `mode` 受控值使用
   */
  onPanelChange?: (value: moment.Moment, mode: MODE) => void;
  /**
   * 一周开始于星期几，默认星期天
   * @defaultValue 0
   */
  startOfWeek?: StartOfWeek;
  /**
   * 点击清除按钮时的回调
   */
  onClear?: () => void;
  /**
   * 输入框失去焦点事件回调
   */
  onBlur?: React.FocusEventHandler;
  /**
   * 是否显示时间选择及相关配置 [sharedTimeProps](#showTime)
   * @defaultValue false
   */
  showTime?: boolean | SharedTimeProps;
  /**
   * 是否显示确认按钮，设置showTime时默认为 true, 其他默认为 false
   */
  needConfirm?: boolean;
  /**
   * 是否显示当前时间按钮，只有在设置 showTime 时才能使用，默认为 true
   */
  showNow?: boolean;
}
// nolint-next-line: long_function
export function DatePicker(props: DatePickerProps) {
  const cssScope = useCssScope('picker');
  const prefixCls = cssScope();
  const locale = useLocaleReceiver('datePicker');
  const defaultSize = useComponentSize();
  const inputGroupSize = useInputGroupSize();
  const {
    style,
    id,
    showTime = false,
    picker = 'date',
    mode = picker,
    disabledDate: disabledDateProp,
    minDate,
    maxDate,
    className,
    allowClear,
    autoFocus,
    disabled,
    // eslint-disable-next-line no-nested-ternary
    placeholder = picker === 'date'
      ? locale.placeholder
      : mode === 'month'
      ? locale.monthPlaceholder
      : locale.yearPlaceholder,
    inputReadOnly,
    format: formatProp,
    size,
    error,
    triggerElement,
    parse,
    onKeyDown = noop,
    onMouseDown = noop,
    onMouseUp = noop,
    onClick = noop,
    suffixIcon = <CalendarOutlined />,
    renderExtraFooter,
    showToday = !showTime && picker === 'date',
    getPopupContainer,
    dropdownClassName,
    popupStyle,
    startOfWeek = 0,
    dateRender,
    onClear,
    onSelect,
    defaultPickerValue,
    defaultOpen = false,
    value: valueProp,
    defaultValue,
    onChange,
    onBlur,
    needConfirm = Boolean(showTime),
    showNow = Boolean(showTime),
    hideWeekNumbers,
    janX,
  } = props;

  const disabledDate = useMemo(() => {
    if (disabledDateProp === undefined) {
      return undefined;
    }

    const cache = new Map<string, boolean>();
    return function (date: moment.Moment) {
      const cacheKey = date.format('YYYYMMDD');
      if (cache.has(cacheKey)) {
        return cache.get(cacheKey) as boolean;
      }
      const isDisabled = disabledDateProp(date);
      cache.set(cacheKey, isDisabled);
      return isDisabled;
    };
  }, [disabledDateProp]);

  const direction = useDirection();
  const Popper = usePopperSidecar();
  const cmpSize = getDefaultSizeAbbr(size || inputGroupSize || defaultSize);
  const showTimeSelection = Boolean(showTime);
  const { formatValue, parseValue, format } = useFormatAndParser({
    mode,
    format: formatProp,
    parse,
    showTime: showTimeSelection,
  });

  warning(!(showNow && showToday), devWarning('datePicker', 'should not set showNow and showToday at same time'));

  const dateToText = React.useCallback((date?: moment.Moment) => formatValue(date), [formatValue]);

  const [value, setValue] = useControlledState<moment.Moment | undefined>(
    // 如果 valueProp 是 moment 形式，采用 valueProp
    // 如果 valueProp 是 undefined 或其他，统一采用 undefined 形式
    moment.isMoment(defaultValue) && defaultValue.isValid() ? defaultValue : undefined,
    moment.isMoment(valueProp) && valueProp.isValid() ? valueProp : undefined,
    {
      onChange,
      isEqual: isDateValueEqual,
      isControlledOverride: 'value' in props,
    },
  );

  /**
   * 当前视图可见的有效date
   */
  const [viewValue, setViewValue] = useSyncState(value, {
    isEqual: isDateValueEqual,
  });

  const validText = dateToText(viewValue);

  /**
   * 输入框的值
   */
  const [text, setText] = useSyncState<string>(validText);

  /**
   * hover相关的值
   */
  const [hoverDate, setHoverDate] = useState<moment.Moment>();

  const hoverText = dateToText(hoverDate);
  const [focused, setFocused] = useState(false);

  const [open, setOpened] = useControlled<boolean>(props, {
    defaultValuePropName: '',
    valuePropName: 'open',
    onChangePropName: 'onOpenChange',
    defaultValue: defaultOpen,
  });

  const panelDivRef = React.useRef<HTMLDivElement>(null);
  const inputDivRef = React.useRef<HTMLDivElement>(null);
  const inputRef = useRef<HTMLInputElement>(null);
  const triggerElementRef = React.useRef<HTMLElement>(null);

  const rootCls = clsx(cssScope(), {
    [cssScope(`--size-${cmpSize}`)]: cmpSize,
    [cssScope('--error')]: error,
    [cssScope('--focus')]: focused,
    [cssScope('--disabled')]: disabled,
  });

  const ctxGetPopupContainer = usePopupPortalContainer();

  const getPickerPopupContainer = (triggerNode: HTMLElement): HTMLElement => {
    const _getPopupContainer = getPopupContainer ?? ctxGetPopupContainer;
    return _getPopupContainer(triggerNode);
  };

  const confirm = useCallback(
    (newValue?: moment.Moment) => {
      if (!isEqual(value, newValue, mode, showTimeSelection)) {
        setValue(newValue);
      }

      // 暂时同步value值，不然受控模式下，可能出现问题
      setViewValue(value);
      setText(dateToText(value));

      setHoverDate(undefined);
      setOpened(false);
    },
    [value, mode, showTimeSelection, setViewValue, setText, dateToText, setHoverDate, setOpened, setValue],
  );

  const cancel = useCallback(() => {
    setOpened(false);
    setViewValue(value);
    setText(dateToText(value));
  }, [dateToText, setOpened, setText, setViewValue, value]);

  function setDateValue(v: SetStateAction<moment.Moment | undefined>, autoCommit = !needConfirm) {
    if (autoCommit) {
      confirm(typeof v === 'function' ? v(viewValue) : v);
    } else {
      setViewValue(v);
    }
  }

  function handleKeyDown(e: React.KeyboardEvent<HTMLInputElement>) {
    onKeyDown(e);

    if (e.key !== 'Tab' && !open) {
      setOpened(true);
      return;
    }

    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    e.keyCode === 13 && confirm(viewValue);
  }

  const handleInputChange: React.ChangeEventHandler<HTMLInputElement> = e => {
    setHoverDate(undefined);
    const editText = e.target.value;
    setText(editText);
    const date = parseValue(editText);
    if (
      date?.isValid() &&
      !isInputDateDisabled(date, {
        minDate,
        maxDate,
        disabledDate,
      })
    ) {
      setViewValue(date);
    }
  };

  const handleClear = () => {
    setValue(undefined);
    // 暂时同步value值，不然受控模式下，可能出现问题
    setViewValue(value);
    setOpened(false);

    onClear?.();
  };

  const handleMouseUp: React.MouseEventHandler<HTMLDivElement> = e => {
    if (inputRef.current) {
      inputRef.current.focus();
    }
    setOpened(true);
    onMouseUp(e);
  };

  const handleFocus: React.FocusEventHandler<HTMLInputElement> = e => {
    setFocused(true);
  };

  const handleBlur: React.FocusEventHandler<HTMLInputElement> = e => {
    onBlur?.(e);
    setFocused(false);
    needConfirm ? cancel() : confirm(viewValue);
  };

  /**
   * 点击弹窗外部行为：
   *    1. showTime
   *       1. confirmButton: 取消
   *       2. 非 confirmButton：提交
   *    2. !showTime
   *        1. confirmButton: 取消
   *        2. 非 confirmButton：取消
   */
  useClickOutsideEffect(
    useMemo(() => [panelDivRef, inputDivRef, triggerElementRef], []),
    open,
    useCallback(() => {
      !needConfirm && showTime ? confirm(viewValue) : cancel();
    }, [cancel, confirm, needConfirm, showTime, viewValue]),
  );

  const [localMode, setLocalMode] = useSyncState<MODE>(mode);

  const triggerElementEnable = Boolean(isFunction(triggerElement) || triggerElement);
  const showAction: PopperShowActionType[] = triggerElementEnable ? ['click'] : [];

  let triggerElementNode: ReactNode;
  if (isFunction(triggerElement)) {
    triggerElementNode = triggerElement({
      text,
      hoverText,
      open,
      value,
    });
  } else {
    triggerElementNode = triggerElement;
  }

  const renderElement = () => {
    if (triggerElementNode) {
      // 如果用户传递了非 ReactElement 这里会直接报错，导致无法正常渲染
      return React.cloneElement(triggerElementNode as React.ReactElement, {
        ref: composeRefs<HTMLElement>(triggerElementRef, (triggerElementNode as any).ref),
      });
    }

    return (
      <div style={style} ref={inputDivRef} className={rootCls}>
        <InputWrapper
          className={clsx(className, `${prefixCls}-dateInput`)}
          showClear={Boolean(viewValue)}
          onClear={handleClear}
          prefixCls={prefixCls}
          onMouseUp={handleMouseUp}
          onMouseDown={onMouseDown}
          onClick={onClick}
          focused={focused}
          suffixIcon={suffixIcon}
          disabled={disabled}
          allowClear={allowClear}
        >
          <NativeInput
            id={id}
            size={getInputSize(format, generateConfig)}
            className={clsx({
              [`${prefixCls}-input`]: true,
              [`${prefixCls}-input-placeholder`]: Boolean(hoverText),
            })}
            placeholder={placeholder}
            disabled={disabled}
            autoFocus={autoFocus}
            ref={inputRef}
            onKeyDown={handleKeyDown}
            onFocus={handleFocus}
            onBlur={handleBlur}
            onChange={handleInputChange}
            readOnly={inputReadOnly || (isFunction(format) && !isFunction(parse))}
            value={hoverText || text}
            data-testid={cssScope('__input')}
          />
        </InputWrapper>
      </div>
    );
  };

  const handleOnSelect = useCallback(
    (newValue?: moment.Moment) => {
      onSelect?.(newValue);
      if (!needConfirm && !showTime) {
        confirm(newValue);
      } else {
        setViewValue(newValue);
      }
    },
    [onSelect, needConfirm, showTime, confirm, setViewValue],
  );

  return (
    <Popper
      destroyOnHide
      popupStyle={popupStyle}
      getPopupContainer={getPickerPopupContainer}
      visible={open}
      action={undefined}
      showAction={showAction}
      hideAction={[]}
      popupClassName={clsx(dropdownClassName, `${prefixCls}-dropdown`)}
      disabled={disabled}
      onVisibleChange={setOpened}
      popup={
        <ListBox className={`${prefixCls}__list__box`} onMouseDown={e => e.preventDefault()} ref={panelDivRef}>
          <DatePickerContextProvider
            value={{
              mode: localMode,
              setMode: setLocalMode,
              value,
              setValue: setDateValue,
              showToday,
              renderExtraFooter,
              setOpened,
              showTime,
              needConfirm,
              showNow,
            }}
          >
            <PickerPanel
              defaultViewDate={defaultPickerValue}
              disabledDate={disabledDate}
              onDateMouseEnter={setHoverDate}
              onContentMouseLeave={() => setHoverDate(undefined)}
              onSelect={handleOnSelect}
              value={viewValue}
              mode={mode}
              startOfWeek={startOfWeek}
              dateRender={dateRender}
              format={format}
              onCancel={cancel}
              onConfirm={() => confirm(viewValue)}
              minDate={minDate}
              maxDate={maxDate}
              hideWeekNumbers={hideWeekNumbers}
              janX={janX}
            />
          </DatePickerContextProvider>
        </ListBox>
      }
      prefixCls={prefixCls}
      popupMotion="slide-up"
      popupPlacement={direction === 'rtl' ? 'bottomRight' : 'bottomLeft'}
      enableRestoreFocus={false}
      initialFocus={null}
      {...{
        showArrow: false,
        arrowHasStroke: false,
        Transition: SlideUpDownTransition,
        enableMaxHeight: false,
      }}
      defaultMainAxisOffsetWithoutArrow={4}
    >
      {renderElement()}
    </Popper>
  );
}
