// @universe-design/react/src/date-picker/index.ts
import { DatePicker as OriginDatePicker } from './DatePicker';
import type { PickerBodyProps } from './panels/picker/Body';
import type { PickerFooterProps } from './panels/picker/Footer';
import type { MonthYearSelectorProps } from './panels/selector/MonthYearSelector';
import { RangePicker as OriginRangePicker } from './RangePicker';

type DatePickerType = typeof OriginDatePicker & {
  RangePicker: typeof OriginRangePicker;
};

const DatePickerCmp = OriginDatePicker as DatePickerType;
DatePickerCmp.RangePicker = OriginRangePicker;

/**
 * Value export
 */
// DatePicker
export const DatePicker = DatePickerCmp;
export { isValidRangeValue } from './utils/date';
export { DatePickerContextProvider } from './Context';
export { PickerPanel as DatePickerPanel } from './panels/picker';
// RangePicker
export { RangePicker, RangePickerProvider } from './RangePicker';
export { RangePickerPanel } from './RangePicker/RangePickerPanel';
export { RangePickerFooter } from './RangePicker/RangePickerFooter';
export { useRangePickerContext } from './RangePicker/Context';

// DateCalendar
export { DateCalendarUI } from './panels/DateCalendar/DateCalendarUI';
export { DateCalendar } from './panels/DateCalendar/DateCalendar';
// Enum
export { Phase, TriggerType } from './RangePicker/types';

/**
 * Type export
 */
export type { PickerPanelProps } from './panels/picker';
export type { DatePickerProps, DatePickerState, DatePickerTriggerElement } from './DatePicker';
export type { DatePickerContextType } from './Context';
export type {
  RangePickerProps,
  RangePickerState,
  RangePickerTriggerElement,
  RangePickerTriggerProps,
  RangePickerContextProps,
  RangePickerPanelProps,
  RangePickerFooterProps,
} from './RangePicker/types';
export type {
  RangePickerSelectionStrategy,
  RangePickerSelectionStrategyContext,
} from './RangePicker/range-picker-strategies';
export type { RangeValue, MODE, PickerType, StartOfWeek, IFormat, IParse } from './interface';
export type { DateCalendarUIProps, InteractivePanelProps } from './panels/DateCalendar/interface';
export type { DateCalendarProps } from './panels/DateCalendar/DateCalendar';

// for docs
export type { MonthYearSelectorProps, PickerBodyProps, PickerFooterProps };
