// @byted/ai-paas-env/dist/index.js
import env from "@byted-service/env";
var src_AI_PAAS_REGION = /*#__PURE__*/ function(AI_PAAS_REGION) {
    AI_PAAS_REGION["LOCALHOST_DIRECT"] = "LOCALHOST_DIRECT";
    AI_PAAS_REGION["LOCALHOST_GATEWAY"] = "LOCALHOST_GATEWAY";
    AI_PAAS_REGION["BOE_SANDBOX"] = "BOE_SANDBOX";
    AI_PAAS_REGION["BOE"] = "BOE";
    AI_PAAS_REGION["CN"] = "CN";
    AI_PAAS_REGION["I18N"] = "I18N";
    AI_PAAS_REGION["BOEI18N"] = "BOEI18N";
    AI_PAAS_REGION["USTTP"] = "USTTP";
    AI_PAAS_REGION["EUTTP"] = "EUTTP";
    AI_PAAS_REGION["I18NBD"] = "I18NBD";
    return AI_PAAS_REGION;
}({});
const BOE_REGIONS = new Set([
    env.VREGION.VREGION_CHINABOE,
    env.VREGION.VREGION_CHINABOE2
]);
const I18N_REGIONS = new Set([
    env.VREGION.VREGION_SINGAPORECENTRAL,
    env.VREGION.VREGION_IDCOMPLIANCE,
    env.VREGION.VREGION_IDCOMPLIANCE2,
    env.VREGION.VREGION_USEAST,
    env.VREGION.VREGION_USEASTBD,
    env.VREGION.VREGION_USCOMPLIANCE,
    env.VREGION.VREGION_EUROPECENTRAL
]);
const I18NBD_REGIONS = new Set([
    env.VREGION.VREGION_ASIASOUTHEASTBD
]);
const BOEI18N_REGIONS = new Set([
    env.VREGION.VREGION_USBOE,
    env.VREGION.VREGION_SINGAPOREBOE
]);
const USTTP_REGIONS = new Set([
    env.VREGION.VREGION_USTTP,
    env.VREGION.VREGION_USTTP2
]);
const EUTTP_REGIONS = new Set([
    env.VREGION.VREGION_EUTTP,
    env.VREGION.VREGION_EUTTP2,
    env.VREGION.VREGION_EUCOMPLIANCE,
    env.VREGION.VREGION_EUCOMPLIANCE2
]);
function inferAIPaaSRegion() {
    const vRegion = env.getCurrentVRegion();
    if (BOE_REGIONS.has(vRegion)) return "BOE";
    if (I18N_REGIONS.has(vRegion)) return "I18N";
    if (I18NBD_REGIONS.has(vRegion)) return "I18NBD";
    if (BOEI18N_REGIONS.has(vRegion)) return "BOEI18N";
    if (USTTP_REGIONS.has(vRegion)) return "USTTP";
    if (EUTTP_REGIONS.has(vRegion)) return "EUTTP";
    return "CN";
}
function ifProdNetwork() {
    return Boolean(process.env.BYTED_HOST_IP) || Boolean(process.env.BYTED_HOST_IPV6);
}
export { src_AI_PAAS_REGION as AI_PAAS_REGION, ifProdNetwork, inferAIPaaSRegion };
