// @universe-design/react/es/notice/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import ErrorFilled from '@universe-design/icons-react/ErrorFilled';
import CloseOutlined from '@universe-design/icons-react/CloseOutlined';
import InfoFilled from '@universe-design/icons-react/InfoFilled';
import WarningFilled from '@universe-design/icons-react/WarningFilled';
import SucceedFilled from '@universe-design/icons-react/SucceedFilled';
import stylex from '@universe-design/stylex';
import { useControlledState } from '@universe-design/react-utilities';
import { Button } from '../button';
import clsx from 'clsx';
import React, { useCallback, useRef } from 'react';
import { resolveSlot, renderSlotNode } from '../shared/slot';
import { devWarning, warning } from '../util';
import { useCssScope } from '../config-provider/hooks';
import { SpinLoadingSvg } from '../_assets/svgs';
import { cssTypographyHeadline } from '../stylex';
import './style';
const iconMap = {
  success: SucceedFilled,
  warning: WarningFilled,
  error: ErrorFilled,
  loading: SpinLoadingSvg,
  info: InfoFilled
};
export const Notice = /*#__PURE__*/React.forwardRef(function (props, ref) {
  const {
    className,
    closable = false,
    align = 'left',
    type = 'info',
    theme = 'filled',
    message,
    description,
    action,
    visible: visibleProp,
    banner = false,
    onClose,
    closeText,
    closeButton,
    afterClose,
    showIcon = true,
    role = 'alert',
    icon,
    prefixCls,
    maxLine,
    ...restProps
  } = props;
  process.env.NODE_ENV !== "production" ? warning(!afterClose, devWarning(Notice.displayName || Notice.name, 'Close motion is not support, so afterClose is not provided.')) : void 0;
  const cssScope = useCssScope('notice', prefixCls);
  const [visible, setVisible] = useControlledState(true, visibleProp);
  const descRef = useRef(null);
  const isNoTitle = message && !description;
  let iconNode = null;
  if (showIcon) {
    if (icon) {
      iconNode = /*#__PURE__*/React.createElement("div", {
        className: cssScope('__statusIcon')
      }, icon);
    }
    if (!iconNode) {
      const StatusIcon = iconMap[type] || iconMap.info;
      iconNode = /*#__PURE__*/React.createElement(StatusIcon, {
        className: cssScope('__statusIcon')
      });
    }
  }
  const isClosable = closeText ? true : closable;
  const handleClick = useCallback(e => {
    setVisible(false);
    onClose && onClose(e);
  }, [setVisible, onClose]);
  const closeBtn = React.useMemo(() => {
    if (closeText) {
      return /*#__PURE__*/React.createElement("span", {
        className: cssScope('__close'),
        onClick: handleClick
      }, closeText);
    }
    const closeButtonSlot = resolveSlot(closeButton, {
      elementType: Button,
      defaultProps: {
        icon: /*#__PURE__*/React.createElement(CloseOutlined, null),
        size: 'small',
        type: 'icon'
      },
      renderByDefault: true
    });
    if (closeButtonSlot) {
      closeButtonSlot.props.className = clsx(cssScope('__close'), closeButtonSlot.props.className);
      const originalOnClick = closeButtonSlot.props.onClick;
      closeButtonSlot.props.onClick = e => {
        originalOnClick === null || originalOnClick === void 0 ? void 0 : originalOnClick(e);
        handleClick(e);
      };
    }
    return renderSlotNode(closeButtonSlot);
  }, [closeButton, closeText, cssScope, handleClick]);
  const innerType = type === 'loading' ? 'info' : type;
  const contentStyle = maxLine ? {
    WebkitLineClamp: maxLine
  } : undefined;
  return visible ? /*#__PURE__*/React.createElement("div", _extends({
    ref: ref,
    className: clsx(cssScope(), cssScope(`-${innerType}`), align === 'center' && isNoTitle && cssScope('-align-center'), closable && cssScope('-closable'), banner && cssScope('-banner'), iconNode && cssScope('-showIcon'), action && cssScope('-has-action'), message && description && cssScope('-has-title'), cssScope(`--${theme}`), className),
    role: role
  }, restProps), iconNode, /*#__PURE__*/React.createElement("div", {
    className: clsx(cssScope('__main'), isNoTitle && cssScope('__main--no-title'))
  }, message && description && /*#__PURE__*/React.createElement("div", {
    className: cssScope('__title')
  }, /*#__PURE__*/React.createElement("div", {
    className: clsx(cssScope('__title__content'), stylex(cssTypographyHeadline))
  }, message), isClosable && closeBtn), (description || isNoTitle) && /*#__PURE__*/React.createElement("div", {
    ref: descRef,
    className: cssScope('__description')
  }, /*#__PURE__*/React.createElement("div", {
    className: cssScope('__description-content'),
    style: contentStyle
  }, description || message), action && /*#__PURE__*/React.createElement("div", {
    className: cssScope('__description-action')
  }, action)), isClosable && isNoTitle && closeBtn)) : null;
});
