// @universe-design/react/es/select/common.js
import escapeRegExp from 'lodash/escapeRegExp';
import { useCssScope } from '../config-provider/hooks';
export function useSelectListCssScope() {
  return useCssScope('select__list');
}
export function useSelectListItemCssScope() {
  return useCssScope('select__list__item');
}
export function useSelectListGroupCssScope() {
  return useCssScope('select__list__group');
}

/**
 * ======== 判断是不是我们提供的 Option 和 OptionGroup 组件 ========
 */
export const UD_SELECT_OPTION_ID = 'UD_SELECT_OPTION';
export const UD_SELECT_OPTION_GROUP_ID = 'UD_SELECT_OPTION_GROUP';
export function isSelectOption(node) {
  var _node$type;
  return node === null || node === void 0 ? void 0 : (_node$type = node.type) === null || _node$type === void 0 ? void 0 : _node$type[UD_SELECT_OPTION_ID];
}
export function isSelectOptionGroup(node) {
  var _node$type2;
  return node === null || node === void 0 ? void 0 : (_node$type2 = node.type) === null || _node$type2 === void 0 ? void 0 : _node$type2[UD_SELECT_OPTION_GROUP_ID];
}
export const defaultFilterOption = (inputValue, option) => {
  const keyword = typeof option.label === 'string' ? option.label : String(option.value);
  return !inputValue || new RegExp(escapeRegExp(inputValue), 'gi').test(keyword);
};
export function getOption(value, optionMap, selectedMap) {
  var _ref, _optionMap$get;
  // FIXME 如果没有在 option map 中找到的话做一个假的 option 返回
  return (_ref = (_optionMap$get = optionMap.get(value)) !== null && _optionMap$get !== void 0 ? _optionMap$get : selectedMap.get(value)) !== null && _ref !== void 0 ? _ref : {
    value
  };
}
export function getReactKey(value) {
  return typeof value === 'boolean' ? String(`${value}-bool`) : value;
}
