// @universe-design/react/es/select/internal.interface.js
/**
 * @public
 */
export let SelectSearchAction = /*#__PURE__*/function (SelectSearchAction) {
  SelectSearchAction[SelectSearchAction["Blur"] = 0] = "Blur";
  SelectSearchAction[SelectSearchAction["Input"] = 1] = "Input";
  SelectSearchAction[SelectSearchAction["AutoClear"] = 2] = "AutoClear";
  return SelectSearchAction;
}({});
export let SelectActiveOptionChangeAction = /*#__PURE__*/function (SelectActiveOptionChangeAction) {
  SelectActiveOptionChangeAction[SelectActiveOptionChangeAction["ArrowUp"] = 0] = "ArrowUp";
  SelectActiveOptionChangeAction[SelectActiveOptionChangeAction["ArrowDown"] = 1] = "ArrowDown";
  SelectActiveOptionChangeAction[SelectActiveOptionChangeAction["MouseMove"] = 2] = "MouseMove";
  SelectActiveOptionChangeAction[SelectActiveOptionChangeAction["Manual"] = 3] = "Manual";
  SelectActiveOptionChangeAction[SelectActiveOptionChangeAction["AfterSearch"] = 4] = "AfterSearch";
  return SelectActiveOptionChangeAction;
}({});

/**
 * @public
 */

/**
 * @public
 * Internal Select 的基础事件处理函数接口
 */

export let SelectActionType = /*#__PURE__*/function (SelectActionType) {
  SelectActionType["Select"] = "Select";
  SelectActionType["DeSelect"] = "DeSelect";
  SelectActionType["Clear"] = "Clear";
  SelectActionType["Set"] = "Set";
  SelectActionType["Toggle"] = "Toggle";
  SelectActionType["Checked"] = "Checked";
  SelectActionType["UnChecked"] = "UnChecked";
  SelectActionType["SelectAll"] = "SelectAll";
  SelectActionType["DeSelectAll"] = "DeSelectAll";
  return SelectActionType;
}({});

// FIXME 这里需要注意一下，现在这样的写法虽然可以正确推断，但是在 mode 或者 labelInValue 缺失的时候会有 onChange 推断为 any function 的问题，需要用户手动指定

// ======================== Multiple Select ========================

export let MultipleSelectTagType = /*#__PURE__*/function (MultipleSelectTagType) {
  MultipleSelectTagType[MultipleSelectTagType["TAG"] = 0] = "TAG";
  MultipleSelectTagType[MultipleSelectTagType["MAX_TAG"] = 1] = "MAX_TAG";
  return MultipleSelectTagType;
}({});

/**
 * @public
 */

/**
 * @public
 */

/**
 * @public
 */

/**
 * @public
 */
