// @universe-design/react/es/select/SelectOption.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.u10tiqix{min-width:0px;}
import urmav4p from "./../.stylex/urmav4p";
import clsx from 'clsx';
import React, { forwardRef } from 'react';
import stylex from '@universe-design/stylex';
import { COMPONENT_VERSION_ID } from '../config-provider';
import { Highlighter } from '../shared';
import { UD_SELECT_OPTION_GROUP_ID, UD_SELECT_OPTION_ID, useSelectListGroupCssScope, useSelectListItemCssScope } from './common';
import { SelectContext } from './context/SelectContext';
import { cssTypographyHeadline } from '../stylex';
import { Text } from '../text/Text';

/**
 * For users
 */
export const SelectOption = () => /*#__PURE__*/React.createElement(React.Fragment, null);
Object.assign(SelectOption, {
  [UD_SELECT_OPTION_ID]: true,
  [COMPONENT_VERSION_ID]: "3.28.3"
});

/**
 * For internal use
 */
export const InternalSelectOption = /*#__PURE__*/forwardRef((_ref, ref) => {
  let {
    value,
    label,
    labelRenderer,
    disabled = false,
    menuItemSelectedIcon,
    title,
    active,
    className,
    // omit
    children,
    ...restProps
  } = _ref;
  const cssScope = useSelectListItemCssScope();
  const {
    valueSet,
    toggle,
    inputValue,
    textOverflow
  } = React.useContext(SelectContext);
  function renderHighligherNode(_value) {
    return /*#__PURE__*/React.createElement(Highlighter, {
      searchWords: [inputValue],
      textToHighlight: _value,
      markRenderer: text => /*#__PURE__*/React.createElement("span", {
        className: clsx(!disabled && cssScope('__content-highlight'))
      }, text)
    });
  }
  const selected = valueSet.has(value);
  let labelNode;
  if (labelRenderer) {
    labelNode = labelRenderer(inputValue);
  } else if (label) {
    labelNode = label;
    if (typeof label === 'string') {
      labelNode = renderHighligherNode(label);
    }
  } else {
    labelNode = renderHighligherNode(String(value));
  }
  return /*#__PURE__*/React.createElement("div", _extends({
    ref: ref,
    className: clsx(className, cssScope(), selected && cssScope('-selected'), active && cssScope('-active'), disabled && cssScope('-disabled')),
    onClick: () => {
      toggle(value);
    },
    title: title
  }, restProps), /*#__PURE__*/React.createElement(Text, {
    as: "div",
    className: cssScope('__content'),
    lineClamp: textOverflow ? 1 : undefined,
    sx: [{
      "props": {
        "Bf4jedk": urmav4p
      }
    }]
  }, labelNode), /*#__PURE__*/React.createElement("div", {
    className: cssScope('__check')
  }, selected ? menuItemSelectedIcon : null));
});

/**
 * For users
 */
export const SelectOptionGroup = () => /*#__PURE__*/React.createElement(React.Fragment, null);
Object.assign(SelectOptionGroup, {
  [UD_SELECT_OPTION_GROUP_ID]: true,
  [COMPONENT_VERSION_ID]: "3.28.3"
});
export const InternalSelectOptionGroup = /*#__PURE__*/forwardRef((_ref2, ref) => {
  let {
    isFirst,
    label,
    children
  } = _ref2;
  const cssScope = useSelectListGroupCssScope();
  return /*#__PURE__*/React.createElement("div", {
    ref: ref,
    className: clsx(cssScope(), isFirst && cssScope('-first'))
  }, /*#__PURE__*/React.createElement("div", {
    className: clsx(cssScope('__title'), stylex(cssTypographyHeadline))
  }, label), children);
});
