// @universe-design/react/es/select/SelectAsync.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React, { forwardRef, useCallback, useEffect, useReducer } from 'react';
import { BehaviorSubject, from, isObservable, of, Subject, throwError, timer } from 'rxjs';
import { switchMap, debounce, withLatestFrom, tap, map, catchError } from 'rxjs/operators';
import { Select } from './Select';
import { useConstant } from '@universe-design/react-utilities';
import { useMountedState, useEffectOnce } from '../hooks';
import { useLoading } from '../loading/useLoading';
import { devWarning, isFunction, isPromise, warning } from '../util';
var Actions = /*#__PURE__*/function (Actions) {
  Actions[Actions["StartFetching"] = 0] = "StartFetching";
  Actions[Actions["GotOptionsSuccessfully"] = 1] = "GotOptionsSuccessfully";
  Actions[Actions["GotOptionsFailed"] = 2] = "GotOptionsFailed";
  return Actions;
}(Actions || {});
export function useSelectAsync(props) {
  const {
    defaultOptions = [],
    loadOptions,
    onSearch,
    debounceTime = 800,
    loadingDelay = 200,
    onBeforeLoadOptions,
    onLoadOptionsError,
    onLoadOptionsSuccess,
    // omit
    options: optionsProp,
    filterOption: filterOptionProp,
    loading: loadingProp,
    ...restProps
  } = props;
  process.env.NODE_ENV !== "production" ? warning(optionsProp === undefined, devWarning('useSelectAsync', 'options will be ignored in async select')) : void 0;
  process.env.NODE_ENV !== "production" ? warning(loadingProp === undefined, devWarning('useSelectAsync', 'loading will be ignored in async select')) : void 0;
  process.env.NODE_ENV !== "production" ? warning(filterOptionProp === undefined, devWarning('useSelectAsync', 'filterOption will be ignored in async select')) : void 0;
  const initialState = useConstant(() => {
    var _restProps$searchValu;
    return {
      loading: false,
      options: Array.isArray(defaultOptions) ? defaultOptions : [],
      searchValue: (_restProps$searchValu = restProps.searchValue) !== null && _restProps$searchValu !== void 0 ? _restProps$searchValu : ''
    };
  });
  const [state, dispatch] = useReducer((_state, action) => {
    switch (action.type) {
      case Actions.StartFetching:
        isFunction(onBeforeLoadOptions) && onBeforeLoadOptions();
        return {
          ..._state,
          loading: true
        };
      case Actions.GotOptionsSuccessfully:
        isFunction(onLoadOptionsSuccess) && onLoadOptionsSuccess(action.payload);
        return {
          ..._state,
          options: action.payload,
          loading: false
        };
      case Actions.GotOptionsFailed:
        isFunction(onLoadOptionsError) && onLoadOptionsError(action.payload);
        return {
          ..._state,
          loading: false,
          error: action.payload
        };
      default:
        throw new Error('Action type not found');
    }
  }, initialState);
  const [delayedLoading] = useLoading({
    loading: state.loading,
    delay: loadingDelay
  });
  const isMounted = useMountedState();
  // const isFirstMount = useFirstMountState();

  // Observable
  const event$ = useConstant(() => new Subject());
  const state$ = useConstant(() => new BehaviorSubject(initialState));
  useEffect(() => {
    const value$ = event$.pipe(debounce(ev => ev.delay ? timer(ev.delay) : of(ev)), withLatestFrom(state$), tap(_ref => {
      let [ev] = _ref;
      if (isMounted() && ev.showLoading) {
        dispatch({
          type: Actions.StartFetching
        });
      }
    }), switchMap(_ref2 => {
      let [ev, _state] = _ref2;
      if (!isFunction(loadOptions)) {
        return of(_state);
      }
      try {
        const ret = loadOptions(ev.searchValue, _state.options);
        const mapOptions = map(_options => ({
          ..._state,
          options: _options
        }));
        // eslint-disable-next-line no-nested-ternary
        return (isObservable(ret) ? ret : isPromise(ret) ? from(ret) : of(ret)).pipe(mapOptions, catchError(err => {
          console.error('Unexpected error occured when loading options: ', err, 'Please handle it manually');
          if (isMounted()) {
            dispatch({
              type: Actions.GotOptionsFailed,
              payload: err
            });
          }
          return of(_state);
        }));
      } catch (err) {
        return throwError(err);
      }
    }));
    const subscription = value$.subscribe(value => {
      if (isMounted()) {
        state$.next(value);
        dispatch({
          type: Actions.GotOptionsSuccessfully,
          payload: value.options
        });
      }
    }, err => {
      console.log('value$ complete with error: ', err);
    });
    return () => {
      subscription.unsubscribe();
    };
  }, [event$, isMounted, loadOptions, state$]); // immutable forever

  useEffectOnce(() => {
    if (defaultOptions === false || Array.isArray(defaultOptions)) {
      return;
    }
    if (isFunction(loadOptions)) {
      event$.next({
        searchValue: initialState.searchValue,
        delay: 0,
        showLoading: false
      });
    }
    return () => {
      state$.complete();
      event$.complete();
    };
  });
  const handleSearch = useCallback((v, meta) => {
    event$.next({
      searchValue: v,
      delay: debounceTime,
      showLoading: true
    });
    isFunction(onSearch) && onSearch(v, meta);
  }, [debounceTime, event$, onSearch]);
  const additionalProps = {
    loading: delayedLoading,
    options: state.options,
    onSearch: handleSearch,
    filterOption: false
  };
  return {
    ...additionalProps,
    ...restProps
  };
}
export const SelectAsync = /*#__PURE__*/forwardRef((props, ref) => {
  const asyncProps = useSelectAsync(props);
  // FIXME error TS2769: No overload matches this call.
  return /*#__PURE__*/React.createElement(Select, _extends({
    ref: ref
  }, asyncProps));
});
