// @universe-design/react/es/select/selectGenerator.js
/* eslint-disable max-lines-per-function */
import DownBoldOutlined from '@universe-design/icons-react/DownBoldOutlined';
import EmptyFilled from '@universe-design/icons-react/EmptyFilled';
import clsx from 'clsx';
import * as React from 'react';
import { useComponentSize, useDirection, useInputGroupSize, useLocale } from '../config-provider/hooks';
import { usePopperSidecar } from '../config-provider/hooks/usePopperSidecar';
import { useSize } from '../hooks';
import { useComplementEvent } from '../hooks/useComplementEvent';
import { SlideUpDownTransition } from '../shared/components/transition/SlideUpDownTransition';
import { compatibleDevWarning, isFunction, isNonNullable, warning } from '../util';
import composeRefs from '../util/composeRefs';
import { Selector } from './Selector';
import { defaultMaxTagRenderer } from './SelectorTag';
import { MultipleSelectorContextProvider } from './context/MultipleSelectorContext';
import { SelectContext } from './context/SelectContext';
import { useSelection } from './context/useSelection';
import './style';
import { usePopupPortalContainer } from '@universe-design/react-provider';
export function selectGenerator(config) {
  const InternalSelect = /*#__PURE__*/React.forwardRef((props, ref) => {
    const locale = useLocale();
    const ctxSize = useComponentSize();
    const inputGroupSize = useInputGroupSize();
    const ctxGetPopupContainer = usePopupPortalContainer();
    const {
      placeholder = locale.global.placeholder,
      allowClear = true,
      style,
      className,
      size = inputGroupSize || ctxSize,
      bordered = 'normal',
      dropdownMatchSelectWidth = false,
      error = false,
      maxTagCount,
      overlayNode,
      getPopupContainer = ctxGetPopupContainer,
      showArrow = true,
      maxTagRenderer = defaultMaxTagRenderer,
      clearIcon = /*#__PURE__*/React.createElement(EmptyFilled, null),
      suffixIcon = /*#__PURE__*/React.createElement(DownBoldOutlined, null),
      dropdownClassName,
      dropdownStyle,
      triggerElement,
      onMouseEnter,
      onMouseLeave,
      multipleSelectorMaxHeight = 188,
      optionFilterProp,
      removeIcon,
      placement = 'bottom',
      forceRender,
      destroyOnHide = false
    } = props;
    process.env.NODE_ENV !== "production" ? warning(!optionFilterProp && !removeIcon, compatibleDevWarning(InternalSelect.displayName || InternalSelect.name, ['optionFilterProp', 'removeIcon'])) : void 0;
    const triggerElementEnable = isNonNullable(triggerElement);
    const [overlayElement, overlayRef] = React.useState(null);
    // selectorRef
    const [selectorContainerRef, {
      width: selectorContainerWidth
    }] = useSize();
    const [selectorContainerElement, selectorContainerStateRef] = React.useState(null);
    const selectCtx = useSelection({
      ...props,
      size
    }, config);
    const {
      focusState,
      overlayVisible,
      focus,
      blur,
      setOverlayVisible,
      dropdownCssScope,
      cssScope,
      inputRef,
      cancelDelayCloseOverlay,
      // onInputFocus,
      setInputting
    } = selectCtx;

    /**
     * 控制 select popup 在点击非内部元素时自动关闭
     */
    useComplementEvent('mousedown', {
      listener: React.useCallback(() => {
        console.log('going to close select popup');
        setOverlayVisible(false);
        // For jest test, in browser input should blur when use click other places than inputElement
        blur();
      }, [blur, setOverlayVisible]),
      // 第一次打开 overlay, window mousedown 触发之前 overlayRef 内容还是空的，所以这里需要使用 ref 来保证正确获取 dom
      elements: React.useMemo(() => [overlayElement, selectorContainerElement], [overlayElement, selectorContainerElement]),
      skipped: !overlayVisible,
      options: {
        /**
         * 防止 mousedown 的元素也监听了 mousedown 且阻止了冒泡
         */
        capture: true
      }
    });
    React.useImperativeHandle(ref, () => ({
      focus,
      blur,
      get input() {
        return inputRef.current;
      }
    }), [blur, focus, inputRef]);

    // ============================= Popup ==============================
    const popupStyle = {
      minWidth: selectorContainerWidth,
      ...(dropdownStyle !== null && dropdownStyle !== void 0 ? dropdownStyle : {})
    };
    if (typeof dropdownMatchSelectWidth === 'number') {
      popupStyle.width = dropdownMatchSelectWidth;
    } else if (dropdownMatchSelectWidth) {
      popupStyle.width = selectorContainerWidth;
    }
    const rootCls = clsx(cssScope(), className, {
      [cssScope('--focus')]: focusState
    });
    const direction = useDirection();
    const Popper = usePopperSidecar();
    let selectorNode = /*#__PURE__*/React.createElement(Selector, {
      showArrow: showArrow,
      allowClear: allowClear,
      maxTagCount: maxTagCount,
      bordered: bordered,
      size: size,
      placeholder: placeholder,
      error: error,
      maxTagRenderer: maxTagRenderer,
      clearIcon: clearIcon,
      suffixIcon: suffixIcon
    });
    if (triggerElementEnable) {
      selectorNode = isFunction(triggerElement) ? triggerElement(selectCtx) : triggerElement !== null && triggerElement !== void 0 ? triggerElement : /*#__PURE__*/React.createElement("div", null);
    }
    const activeTimeoutIdsRef = React.useRef([]);
    React.useEffect(() => () => {
      const activeTimeoutIds = activeTimeoutIdsRef.current;
      activeTimeoutIds.forEach(timeoutId => clearTimeout(timeoutId));
      activeTimeoutIds.splice(0, activeTimeoutIds.length);
    }, []);
    const popupPlacement = React.useMemo(() => {
      if (['top', 'bottom'].includes(placement)) {
        return direction === 'rtl' ? `${placement}Right` : `${placement}Left`; // 目前 right 的 css 样式还没有加
      }
      throw new Error(`Placement ${placement} is not defined.`);
    }, [direction, placement]);
    return /*#__PURE__*/React.createElement("div", {
      ref: composeRefs(selectorContainerRef, selectorContainerStateRef),
      className: rootCls,
      style: style,
      onMouseEnter: onMouseEnter,
      onMouseLeave: onMouseLeave,
      onKeyDown: selectCtx.onInternalKeyDown,
      onKeyUp: selectCtx.onInternalKeyUp,
      onBlur: selectCtx.onContainerBlur,
      onFocus: selectCtx.onContainerFocus,
      onMouseDown: event => {
        setInputting(true);
        const activeTimeoutIds = activeTimeoutIdsRef.current;
        const {
          target
        } = event;
        const popupElement = overlayElement;

        // We should give focus back to selector if clicked item is not focusable
        // 这段代码是为了保证 dropdown 里面有 input 的时候在点击 input 和失焦 input 时都可以正常工作
        if (popupElement && popupElement.contains(target)) {
          const timeoutId = window.setTimeout(() => {
            const index = activeTimeoutIds.indexOf(timeoutId);
            if (index !== -1) {
              // 进入处理阶段，把 timeout id 删掉
              activeTimeoutIds.splice(index, 1);
            }

            // 取消延迟执行的 blur
            cancelDelayCloseOverlay();

            // 如果当前 focus 的不是 popup 内部的元素，就聚焦 selector input
            if (!popupElement.contains(document.activeElement)) {
              var _inputRef$current;
              (_inputRef$current = inputRef.current) === null || _inputRef$current === void 0 ? void 0 : _inputRef$current.focus();
            }
          });
          activeTimeoutIds.push(timeoutId);
        }
      }
    }, /*#__PURE__*/React.createElement(MultipleSelectorContextProvider, {
      value: React.useMemo(() => ({
        maxHeight: multipleSelectorMaxHeight
      }), [multipleSelectorMaxHeight])
    }, /*#__PURE__*/React.createElement(SelectContext.Provider, {
      value: selectCtx
    }, /*#__PURE__*/React.createElement(Popper, {
      getPopupContainer: getPopupContainer,
      forceRender: forceRender,
      destroyOnHide: destroyOnHide,
      action: undefined,
      showAction: [],
      hideAction: [],
      popupPlacement: popupPlacement,
      popupMotion: "slide-up",
      prefixCls: dropdownCssScope(),
      popupClassName: dropdownClassName,
      visible: overlayVisible,
      popupStyle: popupStyle,
      popup: /*#__PURE__*/React.createElement("div", {
        ref: overlayRef
      }, overlayNode),
      enableRestoreFocus: false,
      initialFocus: null,
      showArrow: false,
      arrowHasStroke: false,
      Transition: SlideUpDownTransition,
      enableMaxHeight: false,
      defaultMainAxisOffsetWithoutArrow: 4
    }, selectorNode))));
  });
  if (process.env.NODE_ENV !== "production") {
    InternalSelect.displayName = 'InternalSelect';
  }
  return InternalSelect;
}
