// @universe-design/react/es/select/Selector.js
//.uk73vx1{opacity:0;}
import usfh9mg from "./../.stylex/usfh9mg";
//.u5p0z4x{opacity:1;}
import uxozh73 from "./../.stylex/uxozh73";
//.u1h5fybs{line-height:36px;}
import uh6bwfr from "./../.stylex/uh6bwfr";
//.u336tjw{line-height:28px;}
import u903f20 from "./../.stylex/u903f20";
//.u8xlz6g{line-height:24px;}
import uyhod33 from "./../.stylex/uyhod33";
/* eslint-disable react/destructuring-assignment */
import clsx from 'clsx';
import React, { forwardRef, useContext, useImperativeHandle, useMemo, useRef, useState } from 'react';
import stylex from '@universe-design/stylex';
import { getSizeAbbr } from '../util';
import { isDisabledOption } from './context/useSelection';
import { SelectContext } from './context/SelectContext';
import { getReactKey } from './common';
import { Overflow } from '../overflow';
import { useMultipleSelectorContext } from './context/MultipleSelectorContext';
import NativeInput from '../shared/components/Input';
import { TextOverflow } from '../shared/components/text-overflow';
import { useIsomorphicLayoutEffect as useLayoutEffect, useMouseOver } from '../hooks';
import { EmptyElement } from '../shared/components/FirstLineAlign';
import { useTextOverflowPopperSidecar } from '../config-provider/hooks/useTextOverflowPopperSidecar';
import { sxOverflowEllipsis } from '../stylex';
import { isOverflowed } from '../util/dom';
import { Text } from '../text';
const Input = /*#__PURE__*/forwardRef(function (props, ref) {
  const {
    id,
    inputRef,
    selectorCssScope: cssScope,
    inputValue,
    disabled,
    onInputChange,
    onInputCompositionStart,
    onInputCompositionEnd,
    onInputPaste,
    showSearch,
    tabIndex,
    inputting
  } = useContext(SelectContext);
  useImperativeHandle(ref, () => inputRef.current);
  return /*#__PURE__*/React.createElement(NativeInput, {
    id: id,
    tabIndex: tabIndex,
    role: "combobox",
    autoComplete: "off",
    disabled: disabled,
    type: "search",
    readOnly: !showSearch,
    ref: inputRef,
    className: clsx(cssScope('__search__input'), inputting && cssScope('__search__input-inputting')),
    value: inputValue,
    onChange: onInputChange,
    onPaste: onInputPaste,
    onCompositionStart: onInputCompositionStart,
    onCompositionEnd: onInputCompositionEnd
  });
});
const AutoSizeInput = () => {
  const {
    selectorCssScope: cssScope,
    inputValue
  } = useContext(SelectContext);
  const measureRef = useRef(null);
  const [inputWidth, setInputWidth] = useState(0);
  useLayoutEffect(() => {
    if (measureRef.current) {
      setInputWidth(measureRef.current.offsetWidth);
    }
  }, [inputValue]);
  return /*#__PURE__*/React.createElement("span", {
    className: cssScope('__search__input_container'),
    style: {
      // TODO 不优雅
      width: inputWidth + 2
    }
  }, /*#__PURE__*/React.createElement(Input, null), /*#__PURE__*/React.createElement("span", {
    ref: measureRef,
    className: cssScope('__search__measure'),
    "aria-hidden": true
  }, inputValue, "\xA0"));
};
export function MultipleSelector(_ref) {
  let {
    maxTagCount,
    maxTagRenderer
  } = _ref;
  const {
    valueSet,
    getOption,
    deselect,
    disabled: globalDisabled,
    selectorCssScope: cssScope,
    selectorLabelRenderer,
    tagRenderer
  } = React.useContext(SelectContext);
  const autoSizeInputNode = useMemo(() => /*#__PURE__*/React.createElement(AutoSizeInput, null), []);
  return /*#__PURE__*/React.createElement(Overflow, {
    className: clsx(cssScope('__search')),
    maxCount: maxTagCount,
    suffixClassName: cssScope('__search__suffix'),
    suffixWrap: false,
    keys: [...valueSet].map(getReactKey),
    itemRenderer: opt => {
      const _opt = opt;
      const option = getOption(_opt);
      const disabled = isDisabledOption(option, globalDisabled);
      return tagRenderer({
        title: option.htmlTitle,
        value: option.value,
        disabled,
        closable: !disabled,
        onClose: () => {
          deselect(_opt);
        },
        children: selectorLabelRenderer(_opt)
      });
    },
    restRenderer: omittedKeys => maxTagRenderer({
      disabled: globalDisabled
    }, omittedKeys.map(opt => {
      const _opt = opt;
      const option = getOption(_opt);
      const disabled = isDisabledOption(option, globalDisabled);
      return {
        value: option.value,
        disabled,
        closable: !disabled,
        onClose: () => {
          deselect(_opt);
        },
        children: selectorLabelRenderer(_opt)
      };
    })),
    suffixRenderer: () => autoSizeInputNode
  });
}
export function SingleSelector() {
  const {
    selectorCssScope: cssScope
  } = React.useContext(SelectContext);
  return /*#__PURE__*/React.createElement("div", {
    className: cssScope('__search')
  }, /*#__PURE__*/React.createElement(Input, null));
}
function SelectorForwardRefRenderFunction(_ref2, ref) {
  let {
    size,
    placeholder,
    bordered,
    error,
    maxTagCount,
    allowClear,
    showArrow,
    maxTagRenderer,
    clearIcon,
    suffixIcon
  } = _ref2;
  const {
    multiple,
    focusState,
    overlayVisible,
    clear,
    valueSet,
    inputValue,
    disabled,
    selectorCssScope: cssScope,
    selectorLabelRenderer,
    setOverlayVisible,
    inputRef,
    inputting,
    showSearch
  } = React.useContext(SelectContext);
  const {
    maxHeight: multipleSelectorMaxHeight
  } = useMultipleSelectorContext();
  const {
    isMouseOver,
    handleMouseEnter,
    handleMouseLeave
  } = useMouseOver();
  const TextOverflowPopper = useTextOverflowPopperSidecar();
  const placeholderRef = React.useRef(null);
  const [placeholderTooltipVisible, setPlaceholderTooltipVisible] = React.useState(false);
  let selectorNode;
  if (multiple) {
    selectorNode = /*#__PURE__*/React.createElement(MultipleSelector, {
      maxTagCount: maxTagCount,
      maxTagRenderer: maxTagRenderer
    });
  } else {
    selectorNode = /*#__PURE__*/React.createElement(SingleSelector, null);
  }
  const isSelectValueEmpty = !valueSet.size;
  const textVariant = {
    small: 'body2',
    medium: 'body0',
    large: 'title5'
  }[size];
  // 是否应该显示 placeholder
  const isPlaceholderVisible = isSelectValueEmpty && Boolean(placeholder) && !inputValue;
  /**
   * 这儿加个 key
   * 在切换 placeholder 和 selectItem 时, react 有可能会复用节点
   * 有可能切换到 selectItem 的一瞬间, 由于还是复用的 placeholder 的节点
   * 所以那一瞬间继承了 placeholder 的 class, 导致文字颜色有一瞬间是 placeholder 的颜色
   * 此时如果这个节点上有个 transition 的过渡, 就会导致该节点的颜色闪一下
   */
  let placeholderNode = /*#__PURE__*/React.createElement(Text, {
    key: "placeholder",
    as: "div",
    variant: textVariant,
    className: clsx(cssScope('__placeholder'), stylex(sxOverflowEllipsis)),
    ref: placeholderRef
  }, placeholder);
  if (inputValue) {
    if (multiple && !isSelectValueEmpty) {
      placeholderNode = null;
    } else {
      placeholderNode = /*#__PURE__*/React.createElement("span", {
        className: cssScope('__placeholder')
      });
    }
  } else if (!isSelectValueEmpty) {
    if (multiple) {
      placeholderNode = null;
    } else {
      // first item
      const selectItem = valueSet.keys().next().value;
      placeholderNode = /*#__PURE__*/React.createElement(TextOverflow, {
        key: "selectItem",
        as: "div",
        className: clsx(cssScope('__selectItem'), inputting && cssScope('__selectItem-inputting')),
        lineClamp: 1,
        variant: textVariant,
        disabled: disabled
      }, selectorLabelRenderer(selectItem));
    }
  }
  const showClear = !isSelectValueEmpty && !disabled && allowClear;

  /**
   * 兼容其他 DS
   */

  return /*#__PURE__*/React.createElement("div", {
    ref: ref,
    onMouseEnter: handleMouseEnter,
    onMouseLeave: handleMouseLeave,
    className: clsx(cssScope(), cssScope(`-${getSizeAbbr(size)}`), focusState && cssScope('-focus'), overlayVisible && cssScope('-open'), !showSearch && cssScope('-readOnly'), cssScope(`-border-${function () {
      if (bordered === true) {
        return 'normal';
      } else if (bordered === false) {
        return 'noborder';
      }
      return bordered;
    }()}`), !isSelectValueEmpty && cssScope('-not-empty'), showClear && cssScope('-allowClear'),
    // isMouseOver && showClear && cssScope('-showClear'),
    multiple && cssScope('-multiple'), disabled && cssScope('-disabled'), error && cssScope('-error'), !showArrow && cssScope('-hide-arrow')),
    onClick: _ref3 => {
      let {
        target
      } = _ref3;
      // 这里要用 onClick 不能用 mousedown 和 mouseup，因为 multiple 模式下 tag 的 onClick 做了 stopPropagation
      if (target !== inputRef.current) {
        var _inputRef$current;
        (_inputRef$current = inputRef.current) === null || _inputRef$current === void 0 ? void 0 : _inputRef$current.focus();
      }
      setOverlayVisible();
    },
    style: {
      maxHeight: multiple ? multipleSelectorMaxHeight : undefined
    }
  }, /*#__PURE__*/React.createElement(EmptyElement, {
    sx: [{
      small: {
        "props": {
          "Bg96gwp": uyhod33
        }
      },
      medium: {
        "props": {
          "Bg96gwp": u903f20
        }
      },
      large: {
        "props": {
          "Bg96gwp": uh6bwfr
        }
      }
    }[size]]
  }), /*#__PURE__*/React.createElement(TextOverflowPopper, {
    title: /*#__PURE__*/React.createElement(React.Fragment, null, placeholder),
    checkOverflow: () => placeholderRef.current ? isOverflowed(placeholderRef.current) : false,
    visible: isPlaceholderVisible && placeholderTooltipVisible,
    onVisibleChange: visible => {
      setPlaceholderTooltipVisible(visible);
    }
  }, /*#__PURE__*/React.createElement("div", {
    className: cssScope('__content'),
    onMouseLeave: () => {
      setPlaceholderTooltipVisible(false);
    }
  }, placeholderNode, selectorNode)), showClear ? /*#__PURE__*/React.createElement("div", {
    className: clsx(cssScope('__clear'), stylex(isMouseOver ? {
      "props": {
        "abs64n": uxozh73
      }
    } : {
      "props": {
        "abs64n": usfh9mg
      }
    })),
    onMouseDown: e => {
      e.stopPropagation();
      clear();
    },
    onClick: e => {
      // 防止点击冒泡触发 overlay 打开
      e.stopPropagation();
    }
  }, clearIcon) : null, showArrow ? /*#__PURE__*/React.createElement("div", {
    className: clsx(cssScope('__arrow'))
  }, suffixIcon) : null);
}
// TODO typing is not safe!
export const Selector = /*#__PURE__*/React.forwardRef(SelectorForwardRefRenderFunction);
