// @universe-design/react/es/select/SelectorTag.js
import _extends from "@babel/runtime/helpers/esm/extends";
import clsx from 'clsx';
import React, { useContext } from 'react';
import { Tag } from '../tag';
import { SelectContext } from './context/SelectContext';
import { Popover } from '../popover';
import { Space } from '../space';

/**
 * 默认的 tag renderer
 */
function MultipleSelectTag(_ref) {
  let {
    value,
    title,
    closable,
    disabled,
    onClose,
    children
  } = _ref;
  const {
    selectorCssScope: cssScope,
    size
  } = useContext(SelectContext);
  const className = clsx(cssScope('__tag'));
  return /*#__PURE__*/React.createElement(Tag, {
    key: value,
    title: title,
    className: className,
    color: "neutral-option",
    size: size,
    closable: closable,
    visible: true,
    onClose: onClose,
    disabled: disabled
  }, children);
}
export const defaultTagRenderer = props => /*#__PURE__*/React.createElement(MultipleSelectTag, props);

/**
 * 默认的 max tag renderer
 */
const MultipleSelectMaxTag = _ref2 => {
  let {
    disabled,
    tags
  } = _ref2;
  const {
    selectorCssScope: cssScope,
    size,
    tagRenderer
  } = useContext(SelectContext);
  const className = clsx(cssScope('__tag'));
  const tagNode = /*#__PURE__*/React.createElement(Tag, {
    className: className,
    style: {
      justifyContent: 'center'
    },
    color: "neutral-option",
    size: size,
    disabled: disabled
  }, `+${tags.length}`);
  return /*#__PURE__*/React.createElement(Popover, {
    mode: "pure",
    overlayClassName: cssScope('__overlay'),
    content: /*#__PURE__*/React.createElement("div", {
      className: cssScope('__overlay-scrollArea')
    }, /*#__PURE__*/React.createElement(Space, {
      size: 4,
      wrap: true,
      inline: false,
      shouldItemConfinedToContainer: true,
      itemKey: index => tags[index].value
    }, tags.map(tag => /*#__PURE__*/React.createElement(React.Fragment, {
      key: tag.value
    }, tagRenderer(tag))))),
    visible: tags.length === 0 ? false : undefined // 不指定这个在 tags 从有到无的时候会显示一个空的 popover
    ,
    disabled: tags.length === 0
  }, tagNode);
};
export const defaultMaxTagRenderer = (props, tags) => /*#__PURE__*/React.createElement(MultipleSelectMaxTag, _extends({}, props, {
  tags: tags
}));
