// @universe-design/react/es/select/Select.js
import _extends from "@babel/runtime/helpers/esm/extends";
import ListCheckBoldOutlined from '@universe-design/icons-react/ListCheckBoldOutlined';
import React from 'react';
import { useForceUpdate } from '@universe-design/react-utilities';
import { isNonNullable, toFlatArray } from '../util';
import { isSelectOption, isSelectOptionGroup } from './common';
import { selectGenerator } from './selectGenerator';
import { SelectOverlay } from './SelectOverlay';
const RefSelect = selectGenerator({
  selectorCssScope: 'select',
  dropdownCssScope: 'select',
  defaultOptionLabelProp: option => {
    var _ref, _option$labelRenderer, _option$labelRenderer2;
    return (_ref = (_option$labelRenderer = (_option$labelRenderer2 = option.labelRenderer) === null || _option$labelRenderer2 === void 0 ? void 0 : _option$labelRenderer2.call(option, '')) !== null && _option$labelRenderer !== void 0 ? _option$labelRenderer : option.label) !== null && _ref !== void 0 ? _ref : option.value;
  }
});
function selectChildren2Options(children) {
  return toFlatArray(children).map((child, index) => {
    if (isSelectOption(child)) {
      var _ref2;
      const {
        key,
        props: {
          children: _children,
          ...restProps
        }
      } = child;
      return {
        ...restProps,
        label: _children,
        key: (_ref2 = key !== null && key !== void 0 ? key : restProps.value) !== null && _ref2 !== void 0 ? _ref2 : `select-option-key-${index}`
      };
    } else if (isSelectOptionGroup(child)) {
      const {
        key,
        props: {
          children: _children,
          ...restProps
        }
      } = child;
      return {
        ...restProps,
        key: key || `select-optgroup-key-${index}`,
        options: selectChildren2Options(_children)
      };
    }
    return null;
  }).filter(isNonNullable);
}
export const Select = /*#__PURE__*/React.forwardRef((props, ref) => {
  const {
    options = selectChildren2Options(props.children),
    filterOption = true,
    defaultActiveFirstOption = true,
    menuItemSelectedIcon = /*#__PURE__*/React.createElement(ListCheckBoldOutlined, null),
    listHeight = 248,
    onPopupScroll,
    virtual = true,
    itemHeight = 32,
    filterSort,
    debounceTime,
    showSelectAll,
    onActiveOptionChange,
    ...baseSelectProps
  } = props;
  const internalOptions = React.useMemo(() => options.reduce((pre, cur) => {
    if ('options' in cur) {
      for (const opt of cur.options) {
        pre.push(Object.assign({}, opt, {
          htmlTitle: opt.title
        }));
      }
    } else {
      pre.push(Object.assign({}, cur, {
        htmlTitle: cur.title
      }));
    }
    return pre;
  }, []), [options]);
  const forceUpdate = useForceUpdate();
  // We need force update here since popup dom is render async

  const refSelectRef = React.useRef(null);
  const listRef = React.useRef(null);
  React.useImperativeHandle(ref, () => ({
    get input() {
      var _refSelectRef$current;
      return (_refSelectRef$current = refSelectRef.current) === null || _refSelectRef$current === void 0 ? void 0 : _refSelectRef$current.input;
    },
    focus() {
      var _refSelectRef$current2;
      return (_refSelectRef$current2 = refSelectRef.current) === null || _refSelectRef$current2 === void 0 ? void 0 : _refSelectRef$current2.focus();
    },
    blur() {
      var _refSelectRef$current3;
      return (_refSelectRef$current3 = refSelectRef.current) === null || _refSelectRef$current3 === void 0 ? void 0 : _refSelectRef$current3.blur();
    },
    // deprecated: use scrollInfoView instead
    scrollTo() {
      var _listRef$current;
      for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }
      return (_listRef$current = listRef.current) === null || _listRef$current === void 0 ? void 0 : _listRef$current.scrollIntoView(...args);
    },
    scrollIntoView() {
      var _listRef$current2;
      for (var _len2 = arguments.length, args = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
        args[_key2] = arguments[_key2];
      }
      return (_listRef$current2 = listRef.current) === null || _listRef$current2 === void 0 ? void 0 : _listRef$current2.scrollIntoView(...args);
    },
    setActiveOption(value) {
      var _listRef$current3;
      return (_listRef$current3 = listRef.current) === null || _listRef$current3 === void 0 ? void 0 : _listRef$current3.setActiveOption(value);
    }
  }), []);
  return /*#__PURE__*/React.createElement(RefSelect, _extends({}, baseSelectProps, {
    ref: refSelectRef,
    options: internalOptions,
    overlayNode: /*#__PURE__*/React.createElement(SelectOverlay, {
      ref: listRef,
      filterSort: filterSort,
      defaultActiveFirstOption: defaultActiveFirstOption,
      options: options,
      filterOption: filterOption,
      menuItemSelectedIcon: menuItemSelectedIcon,
      listHeight: listHeight,
      onPopupScroll: onPopupScroll,
      virtual: virtual,
      itemHeight: itemHeight,
      onMouseEnter: function () {
        forceUpdate();
      },
      onActiveOptionChange: onActiveOptionChange,
      debounceTime: debounceTime,
      showSelectAll: showSelectAll
    })
  }));
});
