// @universe-design/react/src/timeline/stories/homeless.stories.tsx
import React from 'react';
import type { TimelineModeType, TimelineLabelPosition } from '../..';
import { Timeline, Space, Radio, Tag, Avatar } from '../..';
import { SizeChange } from '../../_stories/common/sizeChange';
import DownBoldOutlined from '@universe-design/icons-react/DownBoldOutlined';
import MemberFilled from '@universe-design/icons-react/MemberFilled';
import MemberOutlined from '@universe-design/icons-react/MemberOutlined';

export default {
  title: 'Components/Timeline/临时示例',
  component: Timeline,
};

export const BasicUsage = () => {
  const sizeChangeProps = SizeChange.useState();
  const [mode, setMode] = React.useState<TimelineModeType>('left');
  const [lp, setLp] = React.useState<TimelineLabelPosition>('relative');
  const [reverse, seReverse] = React.useState<boolean>(false);
  return (
    <Space style={{ width: '100%' }} direction="vertical" size="large">
      <Space size="large" wrap>
        <SizeChange {...sizeChangeProps} />
        <Space>
          <span>mode：</span>
          <Radio.Group value={mode} onChange={value => setMode(value as TimelineModeType)}>
            <Radio.Button value="left">Left</Radio.Button>
            <Radio.Button value="right">Right</Radio.Button>
            <Radio.Button value="alternate">Alternate</Radio.Button>
          </Radio.Group>
        </Space>
        <Space>
          <span>labelPosition：</span>
          <Radio.Group value={lp} onChange={value => setLp(value as TimelineLabelPosition)}>
            <Radio.Button value="same">same</Radio.Button>
            <Radio.Button value="relative">relative</Radio.Button>
          </Radio.Group>
        </Space>
        <Space>
          <span>reverse</span>
          <Radio.Group value={reverse} onChange={v => seReverse(Boolean(v))}>
            <Radio.Button value={false}>normal</Radio.Button>
            <Radio.Button value={true}>reversed</Radio.Button>
          </Radio.Group>
        </Space>
      </Space>
      <Timeline space={sizeChangeProps.size} mode={mode} labelPosition={lp} reverse={reverse}>
        <Timeline.Item title="事件节点" label="2020-03-05 12:30" description="辅助信息，这里是流程节点的说明" />
        <Timeline.Item title="事件节点" label="2020-03-05 12:30" description="辅助信息，这里是流程节点的说明" />
        <Timeline.Item label="2020-03-05 12:30" description="辅助信息，这里是流程节点的说明" />
        <Timeline.Item
          status="notarrived"
          title="事件节点"
          label="2020-03-05 12:30"
          description="辅助信息，这里是流程节点的说明"
        />
      </Timeline>
      <Timeline space={sizeChangeProps.size} mode={mode} labelPosition={lp} reverse={reverse}>
        <Timeline.Item
          status="done"
          dotType="icon"
          title="事件节点"
          label="2020-03-05 12:30"
          description="辅助信息，这里是流程节点的说明"
        />
        <Timeline.Item
          dotType="icon"
          label="2020-03-05 12:30"
          status="current"
          description="辅助信息，这里是流程节点的说明"
        />
        <Timeline.Item dotType="icon" title="事件节点" status="notarrived" label="2020-03-05 12:30" />
      </Timeline>
      <Timeline space={sizeChangeProps.size} mode={mode} labelPosition={lp} reverse={reverse}>
        <Timeline.Item status="success" dotType="icon" title="成功节点" label="2020-03-05 12:30" />
        <Timeline.Item status="warning" dotType="icon" title="警示/出错节点" label="2020-03-05 12:30" />
        <Timeline.Item status="error" dotType="icon" title="失败节点" label="2020-03-05 12:30" />
      </Timeline>
      <Timeline space={sizeChangeProps.size} mode={mode} labelPosition={lp} reverse={reverse}>
        <Timeline.Item title="事件节点" label="2020-03-05 12:30" description="辅助信息，这里是流程节点的说明" />
        <Timeline.Item label="2020-03-05 12:30" status="current" description="辅助信息，这里是流程节点的说明" />
        <Timeline.Item title="事件节点" status="notarrived" label="2020-03-05 12:30" />
      </Timeline>
      <Timeline space={sizeChangeProps.size} mode={mode} labelPosition={lp} reverse={reverse}>
        <Timeline.Item status="success" title="成功节点" label="2020-03-05 12:30" />
        <Timeline.Item status="warning" title="警示/出错节点" label="2020-03-05 12:30" />
        <Timeline.Item status="error" title="失败节点" label="2020-03-05 12:30" />
      </Timeline>
    </Space>
  );
};

const TimeLabel = (
  <>
    <div>2020-03-05</div>
    <div>12:30</div>
  </>
);

export const Customer = () => {
  const sizeChangeProps = SizeChange.useState();
  const [mode, setMode] = React.useState<TimelineModeType>('left');
  const [lp, setLp] = React.useState<TimelineLabelPosition>('relative');
  const [reverse, seReverse] = React.useState<boolean>(false);
  return (
    <Space style={{ width: '100%' }} direction="vertical" size="large">
      <Space size="large" wrap>
        <SizeChange {...sizeChangeProps} />
        <Space>
          <span>mode：</span>
          <Radio.Group value={mode} onChange={value => setMode(value as TimelineModeType)}>
            <Radio.Button value="left">Left</Radio.Button>
            <Radio.Button value="right">Right</Radio.Button>
            <Radio.Button value="alternate">Alternate</Radio.Button>
          </Radio.Group>
        </Space>
        <Space>
          <span>labelPosition：</span>
          <Radio.Group value={lp} onChange={value => setLp(value as TimelineLabelPosition)}>
            <Radio.Button value="same">same</Radio.Button>
            <Radio.Button value="relative">relative</Radio.Button>
          </Radio.Group>
        </Space>
        <Space>
          <span>reverse</span>
          <Radio.Group value={reverse} onChange={v => seReverse(Boolean(v))}>
            <Radio.Button value={false}>normal</Radio.Button>
            <Radio.Button value={true}>reversed</Radio.Button>
          </Radio.Group>
        </Space>
      </Space>
      <Timeline space={sizeChangeProps.size} mode={mode} labelPosition={lp} reverse={reverse}>
        <Timeline.Item
          status="success"
          title={<div style={{ color: '#34c724' }}>标题颜色可自定义</div>}
          label="2020-03-05 12:30"
          description="辅助信息，这里是流程节点的说明"
        />
        <Timeline.Item
          status="current"
          title="事件节点"
          label="2020-03-05 12:30"
          description={<div style={{ fontSize: '12px', color: '#3370ff' }}>辅助信息颜色可自定义</div>}
        />
        <Timeline.Item
          status="error"
          title="事件节点"
          label="2020-03-05 12:30"
          description="辅助信息，这里是流程节点的说明"
        />
        <Timeline.Item
          status="warning"
          title="事件节点"
          label="2020-03-05 12:30"
          description="辅助信息，这里是流程节点的说明"
        />
        <Timeline.Item
          status="notarrived"
          title="事件节点"
          label="2020-03-05 12:30"
          description="辅助信息，这里是流程节点的说明"
        />
        <Timeline.Item
          status="done"
          dot={<MemberOutlined />}
          title="标签样式可自定义"
          label="2020-03-05 12:30"
          description="自定义标签样式"
        />
      </Timeline>
      <Timeline space={sizeChangeProps.size} mode={mode} labelPosition={lp} reverse={reverse}>
        <Timeline.Item
          status="current"
          label={TimeLabel}
          title={
            <div
              style={{
                marginBottom: '4px',
                display: 'flex',
                alignItems: 'center',
              }}
            >
              节点内容可完全自定义
              <span style={{ color: '#8F959E', fontSize: '12px', margin: '0 12px' }}>次要信息</span>
              <Tag size="small" color="blue" style={{ verticalAlign: 'middle' }}>
                标签
              </Tag>
            </div>
          }
          description="辅助信息，任意自定义内容"
        />
        <Timeline.Item
          status="current"
          label={TimeLabel}
          title={<div style={{ marginBottom: '4px' }}>节点内容可完全自定义</div>}
          description={
            <>
              <div style={{ marginBottom: '4px' }}>
                <Tag
                  style={{ verticalAlign: 'middle' }}
                  shape="round"
                  icon={<Avatar icon={<MemberFilled />} size="extra-small" />}
                >
                  姓名
                </Tag>
                <Tag size="small" color="blue" style={{ verticalAlign: 'middle' }}>
                  标签
                </Tag>
              </div>
              <div>任意自定义内容</div>
            </>
          }
        />
        <Timeline.Item
          status="current"
          label={TimeLabel}
          title={<div style={{ marginBottom: '4px' }}>节点内容可完全自定义</div>}
          description={
            <>
              <Avatar.Group
                shape="circle"
                size="small"
                style={{ marginBottom: '4px' }}
                avatars={[
                  {
                    color: 'blue',
                    src: 'https://people.bytedance.net/cloud/img/42/7edfca5596c142b78886e974420b5f22_p.jpg',
                  },
                  { color: 'carmine', children: 'L' },
                  { color: 'green', icon: <MemberOutlined /> },
                  { color: 'indigo', children: 'V' },
                  { color: 'lime', children: 'E' },
                  { color: 'orange', children: 'U' },
                ]}
              />
              <div>任意自定义内容</div>
            </>
          }
        />
        <Timeline.Item
          status="current"
          label={TimeLabel}
          title={
            <div style={{ marginBottom: '4px' }}>
              <Tag
                style={{ verticalAlign: 'middle' }}
                shape="round"
                icon={<Avatar icon={<MemberFilled />} size="extra-small" />}
              >
                姓名
              </Tag>
              <Tag size="small" color="blue" style={{ verticalAlign: 'middle' }}>
                标签
              </Tag>
            </div>
          }
          description="任意自定义内容"
        />
        <Timeline.Item
          status="current"
          label={TimeLabel}
          description={
            <div style={{ marginBottom: '4px' }}>
              <Tag
                style={{ verticalAlign: 'middle' }}
                shape="round"
                icon={<Avatar icon={<MemberFilled />} size="extra-small" />}
              >
                姓名
              </Tag>
              <Tag size="small" color="blue" style={{ verticalAlign: 'middle' }}>
                标签
              </Tag>
            </div>
          }
        />
      </Timeline>
      <Timeline space={sizeChangeProps.size} mode={mode} labelPosition={lp} pending="Loading" reverse={reverse}>
        <Timeline.Item title="事件节点" label="2020-03-05 12:30" description="辅助信息，这里是流程节点的说明" />
        <Timeline.Item title="事件节点" label="2020-03-05 12:30" description="辅助信息，这里是流程节点的说明" />
        <Timeline.Item label="2020-03-05 12:30" description="辅助信息，这里是流程节点的说明" />
        <Timeline.Item
          status="notarrived"
          title="事件节点"
          label="2020-03-05 12:30"
          description="辅助信息，这里是流程节点的说明"
        />
      </Timeline>
      <Timeline space={sizeChangeProps.size} mode={mode} labelPosition={lp} hangingTail reverse={reverse}>
        <Timeline.Item title="事件节点" label="2020-03-05 12:30" description="辅助信息，这里是流程节点的说明" />
        <Timeline.Item title="事件节点" label="2020-03-05 12:30" description="辅助信息，这里是流程节点的说明" />
        <Timeline.Item
          title="事件节点"
          label="2020-03-05 12:30"
          description={
            <>
              <div>辅助信息，这里是流程节点的说明</div>
              <div
                style={{
                  verticalAlign: 'middle',
                  marginTop: '25px',
                  color: '#646A73',
                }}
              >
                <DownBoldOutlined style={{ marginRight: 4 }} />
                16个节点被收起
              </div>
            </>
          }
        />
      </Timeline>
    </Space>
  );
};
