// @universe-design/react/src/timeline/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Basic from '../demo/0.basic';
import C3Color from '../demo/3.color';
import C4Reverse from '../demo/4.reverse';
import C5Label from '../demo/5.label';
import C6Hanging from '../demo/6.hanging';
import C7Status from '../demo/7.status';
import C8Space from '../demo/8.space';
import CGhostNode from '../demo/ghost-node';
if (!(C0Basic as Story).storyName) {
  (C0Basic as Story).storyName = '0. 基础用法';
  (C0Basic as Story).parameters = {
    docs: {
      description: {
        story: '简单的用法。\n\n',
      },
    },
  };
}
if (!(C3Color as Story).storyName) {
  (C3Color as Story).storyName = '1. 时间轴点颜色';
  (C3Color as Story).parameters = {
    docs: {
      description: {
        story: '设置时间轴点颜色。\n\n',
      },
    },
  };
}
if (!(C4Reverse as Story).storyName) {
  (C4Reverse as Story).storyName = '2. 排序';
  (C4Reverse as Story).parameters = {
    docs: {
      description: {
        story: '设置`reverse`为`true`，可以将所有节点倒序排列。\n\n',
      },
    },
  };
}
if (!(C5Label as Story).storyName) {
  (C5Label as Story).storyName = '5. 标签文本位置';
  (C5Label as Story).parameters = {
    docs: {
      description: {
        story: '通过`mode`设置标签文本位置。\n\n',
      },
    },
  };
}
if (!(C6Hanging as Story).storyName) {
  (C6Hanging as Story).storyName = '6. 孤立轴线';
  (C6Hanging as Story).parameters = {
    docs: {
      description: {
        story: '在最后一个节点展示一条孤立的轴线。\n\n',
      },
    },
  };
}
if (!(C7Status as Story).storyName) {
  (C7Status as Story).storyName = '7. 节点状态';
  (C7Status as Story).parameters = {
    docs: {
      description: {
        story: '通过`status`和`type`可以设置不同的节点状态，每种状态对应了不同的节点颜色或者节点图标。\n\n',
      },
    },
  };
}
if (!(C8Space as Story).storyName) {
  (C8Space as Story).storyName = '8. 节点间距';
  (C8Space as Story).parameters = {
    docs: {
      description: {
        story: '设置节点之间的间距。\n\n',
      },
    },
  };
}
if (!(CGhostNode as Story).storyName) {
  (CGhostNode as Story).storyName = '3. 幽灵节点';
  (CGhostNode as Story).parameters = {
    docs: {
      description: {
        story:
          '当任务状态正在发生，还在记录过程中，可用幽灵节点来表示当前的时间节点。`pending`为真值时，展示幽灵节点，`pending`为 React 元素时，可用于定制该节点内容。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Timeline/官网示例',
  tags: ['autodocs'],
};

export { C0Basic, C3Color, C4Reverse, C5Label, C6Hanging, C7Status, C8Space, CGhostNode };
