// @universe-design/react/src/input-number/index.tsx
import React, { forwardRef, useCallback, useEffect, useImperativeHandle, useRef, useState } from 'react';
import clsx from 'clsx';
import AddOutlined from '@universe-design/icons-react/AddOutlined';
import UpBoldOutlined from '@universe-design/icons-react/UpBoldOutlined';
import DownBoldOutlined from '@universe-design/icons-react/DownBoldOutlined';
import ReduceOutlined from '@universe-design/icons-react/ReduceOutlined';
import { useControlledState } from '@universe-design/react-utilities';
import { getDefaultSizeAbbr, isFunction, isNonNullable, KeyCode } from '../util';
import { useCssScope, useInputGroupSize, useComponentSize } from '../config-provider/hooks';
import noop from 'lodash/noop';
import NativeInput from '../shared/components/Input';
import useSyncState from '../hooks/useSyncState';
import { useCursor } from '@universe-design/react-input';
import {
  MAX_SAFE_INTEGER,
  getPrecision,
  ChangeType,
  getValidValueNumberFromAny,
  internalIsNumber,
  DEFAULT_VALUE,
} from './utils';
import type { InputNumberProps, ValueType } from './interface';
import './style';
// nolint-next-line: long_function
export const InputNumber = forwardRef<HTMLInputElement, InputNumberProps>((props, ref) => {
  const defaultSize = useComponentSize();
  const inputGroupSize = useInputGroupSize();
  const {
    min = -MAX_SAFE_INTEGER,
    max = MAX_SAFE_INTEGER,
    size = inputGroupSize || defaultSize,
    step = 1,
    mode = 'default',
    error,
    readOnly,
    disabled,
    autoFocus,
    precision,
    placeholder,
    onlyNumber = false,
    value: valueInProps,
    defaultValue,
    className,
    style,
    addonAfter,
    addonBefore,
    formatter: formatterProp,
    parser = formatterProp ? (input: string) => input.replace(/[^\w.-]+/g, '') : undefined,
    onChange,
    onPressEnter,
    onBlur,
    onFocus,
    onMouseEnter,
    onMouseLeave,
    onClick,
    onStep,
    controls = true,
  } = props;

  const formatter = formatterProp ?? ((input: string) => input);
  const limitWithMinMax = (value: number) => {
    let val = value;
    if ((max || max === 0) && value > max) {
      val = max;
    }
    if ((min || min === 0) && val < min) {
      val = min;
    }
    return val;
  };

  /**
   * toFixed 方法不能保证准确地做四舍五入，
   * 所以需要四舍五入地做保留小数位必须使用此方法
   */
  const processPrecision = (n: number, currentPrecision: number) => {
    const splitNumber = `${n}`.split('.');
    const integerStr = splitNumber[0] || '0';
    const decimalStr = splitNumber[1] || '0';
    const advancedNum = decimalStr[currentPrecision] ? Number(decimalStr[currentPrecision]) : undefined;
    if (n > 0 && advancedNum && advancedNum >= 5) {
      return n + Number(`0.${'0'.repeat(currentPrecision)}${10 - advancedNum}`);
    }
    if (n < 0 && advancedNum) {
      return Number(`${integerStr}.${decimalStr.slice(0, currentPrecision)}`);
    }
    return n;
  };

  const calculateVal = (rawValue: number | undefined, type: ChangeType) => {
    if (rawValue === undefined || Number.isNaN(rawValue)) {
      if (type === ChangeType.submit) {
        return undefined;
      } else {
        rawValue = DEFAULT_VALUE;
      }
    }
    const rawPrecision = getPrecision(rawValue);
    const minPrecision = getPrecision(min);
    const maxPrecision = getPrecision(max);
    const stepPrecision = getPrecision(step);
    let steppedValue = rawValue;
    switch (type) {
      case ChangeType.addition:
        steppedValue = rawValue + step;
        break;
      case ChangeType.subtraction:
        steppedValue = rawValue - step;
        break;
      default:
        break;
    }
    const result = limitWithMinMax(steppedValue);
    /**
     *  计算 precision
     *  受控的 precision 优先级最高
     *  其次看当前值和 step 的 precision, 取最大值
     *  */
    const currentPrecision = isNonNullable(precision)
      ? precision
      : Math.max(
          ...[
            rawPrecision,
            stepPrecision,
            result === min ? minPrecision : null,
            result === max ? maxPrecision : null,
          ].filter(isNonNullable),
        );
    // 这里最后的 toFixed 只是为了补 0
    return processPrecision(result, currentPrecision).toFixed(currentPrecision);
  };

  const inputRef = useRef<HTMLInputElement>(null);
  const timeout = useRef<ReturnType<typeof setTimeout>>();
  const inputingRef = useRef<boolean>(false);
  const compositionRef = useRef<{
    isComposing: boolean;
    nextValue: string | null;
  }>({
    isComposing: false,
    nextValue: null,
  });
  const [focused, setFocused] = useState<boolean>(Boolean(autoFocus));
  const [handleVisible, setHandleVisible] = useState<boolean>(false);
  const [recordCursor, resetCursor] = useCursor(inputRef, focused);

  const [isError, setError] = useControlledState<boolean>(false, Boolean(error), {
    isControlledOverride: 'error' in props,
  });

  /**
   * @param v 原始值
   * 将原始值转换成 number 类型的 value
   */
  function rawToValidValue(v?: ValueType): number | undefined {
    const toNum = getValidValueNumberFromAny(v);
    if (Number.isNaN(toNum)) {
      return undefined;
    }
    const valueFormatted = calculateVal(toNum, ChangeType.submit);
    return valueFormatted === undefined ? undefined : Number(valueFormatted);
  }

  const [value, setValue] = useControlledState<number | undefined>(
    rawToValidValue(defaultValue),
    rawToValidValue(valueInProps),
    {
      isControlledOverride: 'value' in props,
      onChange,
    },
  );

  /**
   * 将 value 转换成 viewValue (输入框显示的值)
   * @param v
   */
  function valueToViewValue(v?: number) {
    if (v === undefined) {
      return '';
    }
    const valueStr = calculateVal(v, ChangeType.submit);
    return formatter(valueStr ?? '');
  }

  // 输入框显示的值，可能是非法的值或者被 formatter 格式化后的值
  const [viewValue, _setViewValue] = useSyncState<string>(valueToViewValue(value), {
    isEqual(v, preV) {
      const parsed = parser ? parser(v) : v;
      const preParsed = parser ? parser(preV) : preV;
      return rawToValidValue(preParsed) === rawToValidValue(parsed);
    },
  });

  /**
   * composition 参考文档：https://bytedance.larkoffice.com/wiki/XU9ewGUVViCeISkLOVpcvpbNnKd
   * 出问题的老版本微软输入法设置：https://jingyan.baidu.com/article/c1a3101eac4ca19f646deb07.html
   * 会打断组合输入状态的情况：
   * 1. 失去焦点，可能会触发两次 start+end；
   * 2. 切换输入法，老式微软输入法时，中文状态下输入 d，ctrl+shift 会导致 input 展示 $dd（format的值【$d】 + 输入法的值【d】），拼接展示，其他输入法正常
   * 3. 空格/enter 键，相当于选取
   * 4. 直接输入数字，本次重点修复 case，format 赋值不再打断组合输入状态，不会造成输入数字重复的问题
   *  */
  const setViewValue = (v?: number | string) => {
    const input = inputRef.current;
    if (!input) {
      return;
    }
    // format 之后的值
    const nextValue = v === undefined ? '' : v.toString();
    // 组合输入状态时，input 的值与 format 之后的值不一致
    if (compositionRef.current.isComposing && input.value !== nextValue) {
      compositionRef.current.nextValue = nextValue;
      // 将 input 的值改为记录的值，防止打断组合输入状态，end 时再把 format 之后的值赋值
      _setViewValue(input.value);
      return;
    }
    /**
     * case:
     * 1. 非组合状态时
     * 2. 组合状态时，input 的值与 format 之后的值一致（比如: $12 输入 3， input.value 和 format 之后都是 $123）
     */
    compositionRef.current.nextValue = null;
    // 一致时，可以直接把 format 之后的赋值，跳过 onCompositionEnd 的赋值
    _setViewValue(nextValue);
  };

  // 将 viewValue 同步至 value，可能会触发 onChange
  /**
   * @param _viewValue 新提交的值
   * @param type 操作类型
   */
  function commitChange(_viewValue: string, type: ChangeType): number | undefined {
    const parsedValue = parser ? parser(_viewValue.toString()) : _viewValue.toString();

    const flush = !inputingRef.current; // 是否要提交及处理 max/min/precision
    let newIsError = isError;

    if (internalIsNumber(parsedValue) || parsedValue === '') {
      newIsError = false;
    } else if (!onlyNumber) {
      // 如果 parse 出来的值不合法，标记为 error
      newIsError = true;
    }

    const toNum = getValidValueNumberFromAny(parsedValue);
    const valueWithPer = calculateVal(toNum, type);

    let newViewValue: string, newValue: number | undefined;

    if (valueWithPer === undefined) {
      // 判断formatter的参数 -> parsedValue / ""
      newViewValue = inputingRef.current ? formatter(parsedValue) : formatter('');
      newValue = undefined;
    } else {
      newViewValue = formatter(!flush ? parsedValue : valueWithPer);
      newValue = Number(valueWithPer);
    }

    if (flush) {
      newIsError = false;
    }

    if (flush || formatterProp) {
      setViewValue(newViewValue);
    } else {
      setViewValue(_viewValue);
    }

    setValue(newValue);

    if (newIsError !== isError) {
      setError(newIsError);
    }
    return newValue;
  }

  useImperativeHandle<HTMLInputElement | null | undefined, HTMLInputElement | null | undefined>(
    ref,
    () => inputRef.current,
  );

  const stop = useCallback(() => {
    timeout.current && clearTimeout(timeout.current);
  }, []);

  useEffect(() => {
    if (focused && document.activeElement !== inputRef.current) {
      inputRef.current?.focus();
    }
  });

  useEffect(() => () => stop(), [stop]);

  useEffect(() => {
    resetCursor();
  }, [resetCursor]);

  const disableUp = max || max === 0 ? value === max : false;
  const disableDown = min || min === 0 ? value === min : false;

  // css start
  const cssScope = useCssScope('input-number');
  const usedSize = getDefaultSizeAbbr(size);
  const hasAddon = addonAfter || addonBefore;
  const cls = clsx(
    cssScope(),
    cssScope(`--mode-${mode}`),
    {
      [cssScope(`--size-${usedSize}`)]: usedSize,
      [cssScope('--focused')]: focused,
      [cssScope('--disabled')]: disabled,
      [cssScope('--readOnly')]: readOnly,
      [cssScope('--error')]: isError,
      [cssScope('--withoutContorls')]: !controls,
    },
    className,
  );

  const inputCls = clsx(cssScope('__input-wrap'));

  const addonWarp = hasAddon
    ? clsx(
        cssScope('__addon'),
        {
          [cssScope('__addon-before-warp')]: addonBefore,
          [cssScope('__addon-after-warp')]: addonAfter,
          [cssScope('--disabled')]: disabled,
        },
        className,
      )
    : '';

  const handleCls = clsx(cssScope('__handle'), {
    [cssScope('__handle--visible')]: handleVisible,
  });

  const arrowUpCls = clsx(cssScope('__handle-item'), {
    [cssScope('__handle-item--disabled')]: disableUp,
  });

  const arrowDownCls = clsx(cssScope('__handle-item'), {
    [cssScope('__handle-item--disabled')]: disableDown,
  });

  const stepperUpCls = clsx(cssScope('__stepper-btn'), {
    [cssScope('__stepper-btn--disabled')]: disableUp,
  });

  const stepperDownCls = clsx(cssScope('__stepper-btn'), {
    [cssScope('__stepper-btn--disabled')]: disableDown,
  });
  // css end
  const doStep = (inputValue: string, type: ChangeType, recursive?: boolean) => {
    stop();
    inputingRef.current = false;
    const newValue = commitChange(inputValue, type) as number; // 这里可以认为 step 操作出来的结果是个 number
    const typeStr = type === ChangeType.addition ? 'up' : 'down';
    isFunction(onStep) && onStep(newValue, { offset: step, type: typeStr });
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    timeout.current = setTimeout(() => doStep(valueToViewValue(newValue), type, true), recursive ? 200 : 600);
  };

  const handleWrapMouseEnter = (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
    isFunction(onMouseEnter) && onMouseEnter(e);
    setHandleVisible(!disabled && !readOnly);
  };

  const handleWrapMouseLeave = (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
    isFunction(onMouseLeave) && onMouseLeave(e);
    setHandleVisible(false);
  };

  const handleInputKeyUp = (e: React.KeyboardEvent<HTMLInputElement>) => {
    stop();
  };

  const handleInputKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (!inputRef.current) {
      return;
    }

    const inputValue = inputRef.current.value;
    if (e.keyCode === KeyCode.UP) {
      e.preventDefault();
      doStep(inputValue, ChangeType.addition);
    } else if (e.keyCode === KeyCode.DOWN) {
      e.preventDefault();
      doStep(inputValue, ChangeType.subtraction);
    } else if (e.keyCode === KeyCode.ENTER) {
      inputingRef.current = false;
      commitChange(inputValue, ChangeType.submit);
      isFunction(onPressEnter) && onPressEnter(e);
    }
  };

  const handleInputFocus = (e: React.FocusEvent<HTMLInputElement>) => {
    isFunction(onFocus) && onFocus(e);
    setFocused(true);
  };
  const handleInputBlur = (e: React.FocusEvent<HTMLInputElement>) => {
    // 当 inputNumber 放在 Menu 组件里头, 如果按了 Up/Down 键, 会 focus 到 Menu 的其他菜单项
    // 导致没有触发 keyUp 事件, 所以就没触发到 keyUp 里的 stop 方法, 导致 input 中的数字不停自增或自减
    stop();
    isFunction(onBlur) && onBlur(e);
    setFocused(false);
    inputingRef.current = false;
    commitChange(e.target.value, ChangeType.submit);
  };

  const handleUpClick = (e: React.MouseEvent<HTMLSpanElement, MouseEvent>) => {
    /* istanbul ignore else  */
    if (!inputRef.current || disableUp) {
      return;
    }
    e.preventDefault();
    e.stopPropagation();
    if (!focused) {
      setFocused(true);
    }

    doStep(inputRef.current.value, ChangeType.addition);
  };

  const handleDownClick = (e: React.MouseEvent<HTMLSpanElement, MouseEvent>) => {
    if (!inputRef.current || disableDown) {
      return;
    }
    e.preventDefault();
    e.stopPropagation();
    if (!focused) {
      setFocused(true);
    }
    doStep(inputRef.current.value, ChangeType.subtraction);
  };

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    inputingRef.current = true;
    recordCursor();
    // 始终把用户的输入更新到 input 中
    commitChange(e.target.value, ChangeType.submit);
  };

  /**
   * 禁止滚轮事件
   * input[type=number] 时，
   * 有用户填完数字以后有滚动页面页面的操作，当时输入框还没有失焦，所以数字发生了改变
   * 不使用 onWheel 的原因是 https://github.com/facebook/react/pull/19654
   */
  React.useEffect(() => {
    const stopWheel = (e: WheelEvent) => e.preventDefault();
    const input = inputRef.current;
    if (!input) {
      return;
    }
    input.addEventListener('wheel', stopWheel, { passive: false });
    return () => input.removeEventListener('wheel', stopWheel);
  }, []);

  const otherProps = onlyNumber ? { type: 'number' } : {};
  const renderInput = () => (
    <div className={inputCls}>
      <NativeInput
        ref={inputRef}
        placeholder={placeholder}
        value={viewValue}
        onChange={handleChange}
        onKeyUp={disabled ? noop : handleInputKeyUp}
        onKeyDown={disabled || readOnly ? noop : handleInputKeyDown}
        onFocus={disabled || readOnly ? noop : handleInputFocus}
        onBlur={disabled || readOnly ? noop : handleInputBlur}
        autoComplete="off"
        readOnly={readOnly}
        autoFocus={autoFocus}
        disabled={disabled}
        data-testid={cssScope('-input-el')}
        onCompositionStart={() => {
          compositionRef.current.isComposing = true;
        }}
        onCompositionEnd={() => {
          // 组合输入状态结束时，把 format 的值更新到 input 中
          if (compositionRef.current.isComposing && compositionRef.current.nextValue) {
            _setViewValue(compositionRef.current.nextValue);
          }
          compositionRef.current.isComposing = false;
          compositionRef.current.nextValue = null;
        }}
        {...otherProps}
      />
    </div>
  );

  const renderBody = () => {
    if (mode === 'stepper') {
      return (
        <>
          {!readOnly && (
            <div
              className={stepperDownCls}
              onMouseDown={disabled ? noop : handleDownClick}
              onMouseUp={disabled ? noop : stop}
              onMouseLeave={disabled ? noop : stop}
            >
              <ReduceOutlined />
            </div>
          )}
          {renderInput()}
          {!readOnly && (
            <div
              className={stepperUpCls}
              onMouseDown={disabled ? noop : handleUpClick}
              onMouseUp={disabled ? noop : stop}
              onMouseLeave={disabled ? noop : stop}
            >
              <AddOutlined />
            </div>
          )}
        </>
      );
    }

    return (
      <>
        {renderInput()}
        {!readOnly && controls && (
          <div className={handleCls}>
            <span
              className={arrowUpCls}
              onMouseDown={disabled ? noop : handleUpClick}
              onMouseUp={disabled ? noop : stop}
              onMouseLeave={disabled ? noop : stop}
              data-testid="ud__input-number-addition"
            >
              <UpBoldOutlined />
            </span>
            <span
              className={arrowDownCls}
              onMouseDown={disabled ? noop : handleDownClick}
              onMouseUp={disabled ? noop : stop}
              onMouseLeave={disabled ? noop : stop}
              data-testid="ud__input-number-subtraction"
            >
              <DownBoldOutlined />
            </span>
          </div>
        )}
      </>
    );
  };

  const renderAddonBefore = (prefix: string = cssScope()) => {
    if (addonBefore) {
      return <span className={`${prefix}__addon-before`}>{addonBefore}</span>;
    }
    return null;
  };

  const renderAddonAfter = (prefix: string = cssScope()) => {
    if (addonAfter) {
      return <span className={`${prefix}__addon-after`}>{addonAfter}</span>;
    }
    return null;
  };

  if (hasAddon) {
    return (
      <div
        className={addonWarp}
        onMouseEnter={disabled ? noop : handleWrapMouseEnter}
        onMouseLeave={disabled ? noop : handleWrapMouseLeave}
        onClick={onClick}
      >
        {renderAddonBefore()}
        <div className={cls} style={style}>
          {renderBody()}
        </div>
        {renderAddonAfter()}
      </div>
    );
  }

  return (
    <div
      className={cls}
      style={style}
      onMouseEnter={disabled ? noop : handleWrapMouseEnter}
      onMouseLeave={disabled ? noop : handleWrapMouseLeave}
      onClick={onClick}
    >
      {renderBody()}
    </div>
  );
});

export type { InputNumberProps, ValueType };
