// @universe-design/react/src/input-number/utils.ts
import { isString } from '../util';

// eslint-disable-next-line @typescript-eslint/no-magic-numbers
export const MAX_SAFE_INTEGER = Number.MAX_SAFE_INTEGER || 2 ** 53 - 1;

/**
 * 当直接使用 up/down 时
 *  1. 如果设置了 min, 则将值切换到 min
 *  2. 如果没有设置 min, 则将值变为 1
 */
export const DEFAULT_VALUE = 0;

export const internalIsNumber = (val: unknown): val is number => val !== '' && !isNaN(Number(val));

const getNumbersReg = /[.+-]?(?!0\d)\d+(\.\d+)?(E[+-]?\d+)?/gi;

export const getValidValueNumberFromAny = (val: unknown): number => {
  if (internalIsNumber(val)) {
    return Number(val);
  }
  if (isString(val)) {
    const tempRes = val.match(getNumbersReg);
    return tempRes ? Number(tempRes[0]) : NaN;
  }
  return NaN;
};

export const getPrecision = (value: number): number => {
  const valueString = String(value);
  if (valueString.indexOf('e-') >= 0) {
    return parseInt(valueString.slice(valueString.indexOf('e-') + 2), 10);
  }
  let result = 0;
  if (valueString.indexOf('.') >= 0) {
    result = valueString.length - valueString.indexOf('.') - 1;
  }
  return result;
};

export enum ChangeType {
  // 提交： blur 或者 回车
  submit,
  // 使用步进或者 control 的 +1
  addition,
  // 使用步进或者 control 的 -1
  subtraction,
}
