// @universe-design/react/es/date-picker/weekNumberWrapper/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React from 'react';
import clsx from 'clsx';
import { Tooltip } from '../../tooltip';
import { useCssScope, useLocale } from '../../config-provider/hooks';
import './index.less';
const WeekNumberWrapper = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    hideWeekNumbers,
    rowNum,
    colNum,
    baseDate,
    children,
    ...restProps
  } = _ref;
  const cssScope = useCssScope('picker__cell-week-number');
  const wrapperCssScope = useCssScope('picker__wrapper-week-number');
  const headerCssScope = useCssScope('picker__header-week-number');
  const {
    components: {
      datePicker: locale
    }
  } = useLocale();
  if (hideWeekNumbers) {
    return /*#__PURE__*/React.createElement(React.Fragment, null, children);
  }
  return /*#__PURE__*/React.createElement("div", _extends({
    ref: ref,
    className: clsx(wrapperCssScope())
  }, restProps), /*#__PURE__*/React.createElement("div", {
    className: clsx(headerCssScope())
  }, /*#__PURE__*/React.createElement("div", {
    className: clsx(headerCssScope('-name'))
  }), /*#__PURE__*/React.createElement("div", {
    className: clsx(headerCssScope('-content'))
  }, (() => [...new Array(rowNum)].map((_, rowIndex) => {
    const weekEndDate = baseDate.clone().add((rowIndex + 1) * colNum - 1, 'day');
    const weekNumber = weekEndDate.weeks();
    const tooltipTitle = locale.weekNumber.f({
      num: String(weekNumber)
    });
    return /*#__PURE__*/React.createElement("div", {
      className: clsx(cssScope()),
      key: rowIndex
    }, /*#__PURE__*/React.createElement(Tooltip, {
      title: tooltipTitle
    }, /*#__PURE__*/React.createElement("div", {
      className: clsx(cssScope('__inner'))
    }, weekNumber)));
  }))())), /*#__PURE__*/React.createElement("div", {
    className: clsx(headerCssScope('__space'))
  }), children);
});
export default WeekNumberWrapper;
