// @byted-service/logagent/lib/consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultLogger = exports.byteLengthSym = exports.builtinTagKeys = exports.noop = exports.ttlogagentHTTPPort = exports.ttlogagentSocketPath = exports.defaultFlushInterval = exports.sendInterval = exports.maxLen = exports.maxPacketMsgLen = exports.maxPacketLen = void 0;
/**
 * https://code.byted.org/log_market/ttlogagent_gosdk/blob/master/sdk.go#L50
 *
 * unix socket buffer size is: 160K on 32-bit linux, 208K on 64-bit linux. packetSizeLimit must be less than above.
 */
exports.maxPacketLen = 120 << 10;
exports.maxPacketMsgLen = 100 << 10;
// TODO: 支持异步流式处理
exports.maxLen = 64 * 1024;
exports.sendInterval = 10e3;
exports.defaultFlushInterval = 10e3;
exports.ttlogagentSocketPath = '/opt/tmp/ttlogagent/unixstream.sock';
exports.ttlogagentHTTPPort = 29503;
const noop = () => {
    //
};
exports.noop = noop;
exports.builtinTagKeys = [
    '_taskName',
    '_language',
    '_process',
    '_level',
    '_ts',
    '_location',
    '_logid',
    '_spanID',
    '_psm',
    '_cluster',
    '_idc',
    '_deployStage',
    '_host',
    '_podName',
];
exports.byteLengthSym = Symbol.for('#byteLength');
exports.defaultLogger = {
    fatal: console.log.bind(console),
    error: console.log.bind(console),
    warn: console.log.bind(console),
    info: console.log.bind(console),
    debug: console.log.bind(console),
    trace: console.log.bind(console),
};
