// @byted-service/logagent/lib/helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.truncate = exports.getMessageSender = exports.getGateway = exports.sendViaUnixDomainSocket = exports.sendViaHTTPProtocol = void 0;
const tslib_1 = require("tslib");
const net_1 = tslib_1.__importDefault(require("net"));
const fs_1 = require("fs");
const child_process_1 = require("child_process");
const node_fetch_1 = tslib_1.__importDefault(require("node-fetch"));
const env_1 = require("@byted-service/env");
const pool_1 = require("./pool");
const consts_1 = require("./consts");
const sendViaHTTPProtocol = (host, port) => async function (messages) {
    // 由于这里的 json 格式是固定的，所以直接使用字符串拼接，以提升性能
    const payload = `{"data":[${messages
        .map((message) => `[${message.data.join(',')}]`)
        .join(',')}]}`;
    try {
        await (0, node_fetch_1.default)(`http://${host}:${port}`, {
            method: 'PUT',
            body: payload,
            headers: {
                'content-type': 'application/json',
            },
        });
        /**
         * 整体代码重构会有大量改动，所以这里先暂时用 this 来解决
         */
        this.metrics.success(messages);
    }
    catch (err) {
        /**
         * 整体代码重构会有大量改动，所以这里先暂时用 this 来解决
         */
        this.metrics.sendHttpError(messages);
    }
};
exports.sendViaHTTPProtocol = sendViaHTTPProtocol;
const sendViaUnixDomainSocket = (socketPath, logger) => async function (messages) {
    const successMessages = [];
    const failedMessages = [];
    // TODO pipe socket失效的期间可以考虑改用http发送日志？
    const sock = await pool_1.pool.acquire({ socketPath, logger });
    for (let i = 0; i < messages.length; i++) {
        const buf = messages[i].data;
        const header = Buffer.alloc(8);
        header.writeInt32BE(0x10100000, 0);
        header.writeInt32BE(buf.length, 4);
        try {
            sock.write(Buffer.concat([header, buf]));
            successMessages.push(messages[i]);
        }
        catch (err) {
            /* istanbul ignore next */
            failedMessages.push(messages[i]);
        }
    }
    pool_1.pool.release(sock);
    if (successMessages.length > 0) {
        /**
         * 整体代码重构会有大量改动，所以这里先暂时用 this 来解决
         */
        this.metrics.success(successMessages);
    }
    /* istanbul ignore next */
    if (failedMessages.length > 0) {
        /**
         * 整体代码重构会有大量改动，所以这里先暂时用 this 来解决
         */
        this.metrics.sendSocketError(failedMessages);
    }
};
exports.sendViaUnixDomainSocket = sendViaUnixDomainSocket;
// TODO 下掉这一整块功能
/* istanbul ignore next */
const getGateway = () => {
    // gulu bootstrap.sh 里面定义了 ROUTE_IP
    if (process.env.ROUTE_IP) {
        return process.env.ROUTE_IP;
    }
    if ((0, env_1.isInTCE)()) {
        if (process.env.IS_HOST_NETWORK) {
            return (0, env_1.isIPv6Only)() ? '::1' : '127.0.0.1';
        }
        // TCE 实例上，优先使用 MY_HOST_IP 获取，`ip route` 的方式不准确
        if (process.env.MY_HOST_IPV6) {
            return process.env.MY_HOST_IPV6;
        }
        if (process.env.MY_HOST_IP) {
            return process.env.MY_HOST_IP;
        }
        const pattern = /default via (\S+)/;
        try {
            const route = ((0, child_process_1.execSync)('/opt/tiger/consul_deploy/bin/determine_dc.sh', {
                encoding: 'utf-8',
            }) || '').trim();
            const match = route.match(pattern);
            if (match) {
                return match[1];
            }
        }
        catch (err) {
            // IGNORE
        }
    }
    return (0, env_1.isIPv6Only)() ? '::1' : '127.0.0.1';
};
exports.getGateway = getGateway;
/**
 * getMessageSender()返回发送日志的方法。
 *
 * 如果当前环境支持logagent的unix domain socket，则logagent会优先利用这种方式发送数据；否则
 * 会通过http发送数据。
 */
function getMessageSender(options) {
    const gateway = (0, exports.getGateway)();
    let socketPath = '';
    let host = '';
    let port = 0;
    switch (true) {
        case typeof options.socketPath === 'string':
            socketPath = options.socketPath;
            return (0, exports.sendViaUnixDomainSocket)(socketPath, options.logger);
        case typeof options.host === 'string' &&
            typeof options.port === 'number':
            host = options.host;
            host = net_1.default.isIPv6(host) ? `[${host}]` : host;
            port = options.port;
            return (0, exports.sendViaHTTPProtocol)(host, port);
        case (0, fs_1.existsSync)(consts_1.ttlogagentSocketPath):
            socketPath = consts_1.ttlogagentSocketPath;
            return (0, exports.sendViaUnixDomainSocket)(socketPath, options.logger);
        default:
            host = net_1.default.isIPv6(gateway) ? `[${gateway}]` : gateway;
            port = consts_1.ttlogagentHTTPPort;
            return (0, exports.sendViaHTTPProtocol)(host, port);
    }
}
exports.getMessageSender = getMessageSender;
const truncateMark = Buffer.from(' [ truncated by @byted-service/logagent');
const truncate = (data, len) => {
    if (len < truncateMark.byteLength) {
        return new Uint8Array();
    }
    if (data.byteLength <= len) {
        return data;
    }
    const newData = data.slice(0, len);
    newData.set(truncateMark, len - truncateMark.byteLength);
    return newData;
};
exports.truncate = truncate;
