// @byted-service/logagent/lib/logagent.js
"use strict";
/**
 * 文档地址：https://docs.bytedance.net/doc/lQkZlmuWeu7fZKVXlG8cyh
 * 参考代码：https://code.byted.org/log_market/gosdk
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.LogAgent = void 0;
const tslib_1 = require("tslib");
const env_1 = require("@byted-service/env");
const metrics_1 = tslib_1.__importDefault(require("@byted-service/metrics"));
const helper_1 = require("./helper");
const metrics_2 = require("./metrics");
const protocol_1 = require("./protocol");
const consts_1 = require("./consts");
class LogAgent {
    constructor(options = {}) {
        this.psm = options.psm || (0, env_1.getPSM)();
        this.sendInterval = options.sendInterval || consts_1.sendInterval;
        this.maxLen = options.maxLen || consts_1.maxLen;
        this.logger = options.logger || consts_1.defaultLogger;
        this.sendBuffer = (0, helper_1.getMessageSender)(Object.assign(Object.assign({}, options), { logger: this.logger }));
        this.packLen = 0;
        this.messages = [];
        this.nextSendTime = Date.now() + this.sendInterval;
        this.timer = undefined;
        this.metrics = (0, metrics_2.wrapMetrics)(options.metrics ||
            new metrics_1.default({
                defaultPrefix: 'toutiao.infra.ttlogagent.nodejssdk_v1',
                logger: this.logger,
            }));
    }
    /**
     * 立即发送所有待发送的日志
     */
    flush() {
        this.nextSendTime = Date.now() + this.sendInterval;
        if (!this.messages.length) {
            return;
        }
        this.sendBuffer(this.messages).catch(consts_1.noop);
        this.messages = [];
        this.packLen = 0;
    }
    /**
     * 检查是当前时间点是否应该发送待发送的日志；如果没有发现定时器的话，会创建一个定时器来发送日志
     */
    checkTimeout() {
        const now = Date.now();
        if (now >= this.nextSendTime) {
            this.flush();
        }
        else if (!this.timer) {
            this.timer = setTimeout(() => {
                this.timer = undefined;
                if (this.messages.length) {
                    this.checkTimeout();
                }
            }, this.nextSendTime - now).unref();
        }
    }
    /**
     * 可以通过两种方式来调用接口，如果只需要指定taskName，则：
     *
     * ```js
     * logagent.send('MESSAGE_CONTENT_HERE')
     * ```
     *
     * 如果定义tag，可以传入一个LogAgentMessageType结构的数据：
     *
     * ```js
     * logagent.send({
     *   msg: 'MESSAGE_CONTENT_HERE',
     *   tags: {
     *     _level: 'info',
     *   }
     * })
     * ```
     *
     * @param taskName
     * @param msg
     */
    send(taskName, msg) {
        if (typeof taskName === 'object' && taskName.msg) {
            msg = taskName;
            taskName = this.psm;
        }
        const message = this._encode(taskName, msg);
        this.packLen += message.data.length;
        this.messages.push(message);
        if (this.packLen >= this.maxLen) {
            this.flush();
        }
        this.checkTimeout();
    }
    /**
     * @private
     * @param taskName
     * @param originMsg
     */
    _encode(taskName, originMsg) {
        /**
         * https://docs.bytedance.net/doc/uhnP66an8j3qU0mn2Kjywd
         * https://code.byted.org/log_market/ttlogagent/blob/master/agent/formatter/text_formatter.go#L110
         *
         * 数据格式：model.Msg
         * 最终结果：_level _ts _location _host _psm _logid _cluster _deployStage _idc _spanID ...tags _msg=日志内容
         */
        const tags = Object.assign({ _taskName: taskName, _level: 'info', _ts: Date.now(), _location: 'byted-service/logagent.js:0', _host: (0, env_1.getIP)(), _psm: this.psm, _logid: '-', _cluster: (0, env_1.getCluster)(), _deployStage: (0, env_1.getStage)(), _idc: (0, env_1.getIDC)(), _spanID: 0, _language: 'js', _podName: (0, env_1.getPodName)(), _process: process.pid }, originMsg.tags);
        const message = { originMsg, tags, data: new Uint8Array() };
        try {
            const msgBuf = Buffer.from(originMsg.msg);
            message.data = (0, protocol_1.encode)({
                msg: (0, helper_1.truncate)(msgBuf, consts_1.maxPacketMsgLen),
                tags: Object.keys(tags).map((key) => ({
                    key,
                    value: String(tags[key]),
                })),
            });
            /**
             * 由于当前版本并未限制 tags 的长度，所以可能出现，tags 过多导致整体长度过长的情况
             * 所以这里优先保证 tags 的完整性
             */
            if (message.data.byteLength > consts_1.maxPacketLen) {
                message.data = (0, protocol_1.encode)({
                    msg: (0, helper_1.truncate)(msgBuf, Math.min(msgBuf.byteLength, consts_1.maxPacketMsgLen) -
                        (message.data.byteLength - consts_1.maxPacketLen)),
                    tags: Object.keys(tags).map((key) => ({
                        key,
                        value: String(tags[key]),
                    })),
                });
                /**
                 * 就算再次截断处理了，也可能会出现光 tags 就已经超出长度限制的情况，没啥办法只能打个 metrics
                 */
                if (message.data.byteLength > consts_1.maxPacketLen) {
                    this.metrics.packetTooLongError([message]);
                }
            }
        }
        catch (err) {
            this.metrics.encodeError([message]);
            throw err;
        }
        return message;
    }
}
exports.LogAgent = LogAgent;
