// @byted-service/logagent/lib/metrics_v2.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.wrapMetrics = void 0;
const wrapMetrics = (metrics, logagent) => {
    const key = `${logagent.psm}.send`;
    return {
        success(msgs) {
            metrics.emitCounter(key, msgs.length, '', { status: 'success' });
        },
        encodeError() {
            metrics.emitCounter(key, 1, '', {
                status: 'error',
                reason: 'encodeError',
            });
        },
        packetTooLongError() {
            metrics.emitCounter(key, 1, '', {
                status: 'error',
                reason: 'packetTooLongError',
            });
        },
        sendHttpError(msgs) {
            metrics.emitCounter(key, msgs.length, '', {
                status: 'error',
                reason: 'sendHttpError',
            });
        },
        sendSocketError(msgs) {
            metrics.emitCounter(key, msgs.length, '', {
                status: 'error',
                reason: 'sendSocketError',
            });
        },
    };
};
exports.wrapMetrics = wrapMetrics;
