// @byted-service/logagent/lib/metrics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.wrapMetrics = void 0;
const tslib_1 = require("tslib");
const lodash_groupby_1 = tslib_1.__importDefault(require("lodash.groupby"));
const _ = {
    groupBy: lodash_groupby_1.default,
};
const wrapMetrics = (metrics) => {
    return {
        emitMetrics(msgs, tags) {
            const msgMap = _.groupBy(msgs, 'tags._taskName');
            for (const taskName of Object.keys(msgMap)) {
                metrics.emitCounter(`${taskName === 'undefined' ? 'unknow' : taskName}.send`, msgMap[taskName].length, '', tags);
            }
        },
        success(msgs) {
            this.emitMetrics(msgs, { status: 'success' });
        },
        encodeError(msgs) {
            this.emitMetrics(msgs, {
                status: 'error',
                reason: 'encodeError',
            });
        },
        packetTooLongError(msgs) {
            this.emitMetrics(msgs, {
                status: 'error',
                reason: 'packetTooLongError',
            });
        },
        sendHttpError(msgs) {
            this.emitMetrics(msgs, {
                status: 'error',
                reason: 'sendHttpError',
            });
        },
        sendSocketError(msgs) {
            this.emitMetrics(msgs, {
                status: 'error',
                reason: 'sendSocketError',
            });
        },
    };
};
exports.wrapMetrics = wrapMetrics;
