// @byted-service/logagent/lib/pool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pool = exports.poolCache = exports.idCreator = void 0;
const netpool_1 = require("@byted-service/netpool");
const idSym = Symbol.for('id');
exports.idCreator = (0, netpool_1.genIdCreator)('@byted-service/logagent');
exports.poolCache = {
    cache: new Map(),
    get(options) {
        let cacheItem = this.cache.get(options.socketPath);
        if (!cacheItem) {
            cacheItem = new Map();
        }
        let pool = cacheItem.get(options.logger);
        if (!pool) {
            pool = new netpool_1.TCPPool({
                id: (0, exports.idCreator)(),
                logger: options.logger,
                connectOptions: {
                    path: options.socketPath,
                },
                create: (socket) => socket.unref(),
            });
            pool[idSym] = options;
            cacheItem.set(options.logger, pool);
            this.cache.set(options.socketPath, cacheItem);
        }
        return pool;
    },
    values() {
        const pools = [];
        for (const [, cacheItem] of this.cache) {
            for (const [, pool] of cacheItem) {
                pools.push(pool);
            }
        }
        return pools;
    },
    delete(pool) {
        const { socketPath, logger } = pool[idSym];
        const cacheItem = this.cache.get(socketPath);
        if (cacheItem) {
            cacheItem.delete(logger);
            if (cacheItem.size === 0) {
                this.cache.delete(socketPath);
            }
        }
    },
};
exports.pool = new netpool_1.DynamicPool({
    poolCache: exports.poolCache,
});
