// @universe-design/react/src/notification/stories/homeless.stories.tsx
import type { ReactNode } from 'react';
import React from 'react';
import type { Meta } from '@storybook/react';
import { Button, Space, notification, toast, Divider } from '../..';
import MoreCloseOutlined from '@universe-design/icons-react/MoreCloseOutlined';
import './index.less';

export default {
  title: 'Components/Notification/临时示例',
  component: undefined,
} satisfies Meta;

const overlengthTitle = '这是超长的标题';
const overlengthStr = '这是条消息，这是条消息，这是条消息，这是条消息，这是条消息。';

export const DifferentStates = (): ReactNode => (
  <div>
    <h2>类型</h2>
    <Button
      onClick={() =>
        notification.success({
          message: '成功，不自动关闭',
          description: overlengthStr,
          duration: null,
        })
      }
    >
      成功
    </Button>
    <Button
      onClick={() =>
        notification.info({
          message: '错误，5s 自动关闭',
          description: overlengthStr,
          duration: 5000,
        })
      }
    >
      提醒
    </Button>
    <Button
      onClick={() =>
        notification.warning({
          message: '错误，10s 自动关闭',
          description: overlengthStr,
          duration: 10000,
        })
      }
    >
      警告
    </Button>
    <Button
      onClick={() =>
        notification.error({
          message: '错误，15s 自动关闭',
          description: overlengthStr,
          duration: 15000,
        })
      }
    >
      错误
    </Button>

    <h2>位置</h2>
    <Button
      onClick={() =>
        notification.info({
          message: '提示',
          description: overlengthStr,
          placement: 'topLeft',
        })
      }
    >
      左上
    </Button>
    <Button
      onClick={() =>
        notification.info({
          message: '提示',
          description: overlengthStr,
          placement: 'topRight',
        })
      }
    >
      右上
    </Button>
    <Button
      onClick={() =>
        notification.info({
          message: '提示',
          description: overlengthStr,
          placement: 'bottomLeft',
        })
      }
    >
      左下
    </Button>
    <Button
      onClick={() =>
        notification.info({
          message: '提示',
          description: overlengthStr,
          placement: 'bottomRight',
        })
      }
    >
      右下
    </Button>

    <h2>超长内容</h2>
    <Button
      onClick={() => {
        notification.info({
          message: overlengthTitle + overlengthTitle + overlengthTitle + overlengthTitle,
          description: overlengthStr + overlengthStr + overlengthStr,
          placement: 'topLeft',
        });
      }}
    >
       超长内容
    </Button>

    <h2>更多配置</h2>
    <Divider text="隐藏 Icon" />
    <Button
      onClick={() =>
        notification.info({
          message: '提示',
          description: overlengthStr,
          showIcon: false,
        })
      }
    >
      隐藏 icon
    </Button>
    <Divider text="替换 icon" />
    <Button
      onClick={() => {
        notification.error({
          message: '提醒',
          description: '出错了！点击查看',
          icon: <MoreCloseOutlined />,
        });
      }}
    >
      点我
    </Button>
    <Button
      onClick={() => {
        notification.error({
          message: '提醒',
          placement: 'bottomRight',
          description: '出错了！点击查看',
          icon: <MoreCloseOutlined />,
        });
      }}
    >
      点我
    </Button>
    <Button
      onClick={() => {
        notification.error({
          message: '提醒',
          placement: 'topLeft',
          description: '出错了！点击查看',
          icon: <MoreCloseOutlined />,
        });
      }}
    >
      点我
    </Button>
    <Divider text="手动关闭" />
    <Button
      onClick={() => {
        const key = `error-notice-${new Date().getTime()}`;
        notification.error({
          key,
          message: '警告',
          description: '出错了！请点击查看 (点击内容才会关闭)',
          onClick: () => notification.close(key),
          onClose: () => toast.success('notification 关闭了'),
        });
      }}
    >
      点我
    </Button>
    <Divider text="关闭全部打开的 Notification" />
    <Button
      onClick={() =>
        notification.info({
          message: '提示',
          description: overlengthStr,
        })
      }
    >
      点我打开 Notification
    </Button>
    <Button type="primary" onClick={() => notification.destroy()}>
      点我关闭所有 Notification
    </Button>
    <Divider text="添加事件监听" />
    <Button
      onClick={() => {
        notification.info({
          message: '添加事件监听',
          description: overlengthStr,
          onClick: () => toast.success('内容被点击了！'),
          onClose: () => toast.success('关闭了！'),
        });
      }}
    >
      点我
    </Button>
    <Divider text="添加自定义按钮" />
    <Button
      onClick={() => {
        const key = `customer-notice-${new Date().getTime()}`;
        notification.info({
          key,
          message: '添加事件监听',
          description: overlengthStr,
          btn: () => (
            <>
              <Button onClick={() => notification.close(key)}>取消</Button>
              <Button onClick={() => notification.close(key)} type="primary">
                知道了
              </Button>
            </>
          ),
        });
      }}
    >
      点我
    </Button>
    <Button
      onClick={() => {
        const key = `customer-notice-${new Date().getTime()}`;
        notification.info({
          key,
          message: '添加事件监听',
          description: overlengthStr,
          duration: null,
          btn: () => (
            <div style={{ textAlign: 'right' }}>
              <Button onClick={() => notification.close(key)}>取消</Button>
              <Button className="notification-btn-margin-right" onClick={() => notification.close(key)} type="primary">
                知道了
              </Button>
            </div>
          ),
        });
      }}
    >
      点我 (按钮右对齐)
    </Button>
    <Button
      onClick={() => {
        const key = `customer-notice-${new Date().getTime()}`;
        notification.open({
          key,
          message: '添加事件监听',
          description: overlengthStr,
          duration: 0,
          btn: () => (
            <>
              <Button onClick={() => notification.close(key)}>取消</Button>
              <Button className="notification-btn-margin-right" onClick={() => notification.close(key)} type="primary">
                知道了
              </Button>
            </>
          ),
        });
      }}
    >
      点我 (按钮右对齐)
    </Button>

    <Divider text="完全自定义" />
    <Button
      onClick={() => {
        const key = `customer-notice-${new Date().getTime()}`;
        notification.open({
          key,
          message: '标题',
          icon: <MoreCloseOutlined />,
          style: { marginTop: '10vh' },
          description: '自定义 icon、关闭时间（10s）、点击内容会提前关闭',
          onClick: () => {
            notification.close(key);
            toast.success('自定义的 notification 被手动提前关闭了');
          },
          onClose: () => toast.success('自定义的 notification 被关闭了'),
        });
      }}
    >
      点我
    </Button>

    <Divider text="使用 config 定义全局设置" />
    <Button onClick={() => notification.config({ placement: 'topLeft' })}>设置打开位置 - 左上</Button>
    <Button onClick={() => notification.config({ placement: 'bottomLeft' })}>设置打开位置 - 左下</Button>
    <Button onClick={() => notification.config({ duration: 5000 })}>设置自动关闭时间 - 5s</Button>
    <Button onClick={() => notification.config({ top: 100 })}>设置距离顶部 100px</Button>
    <Button onClick={() => notification.config({ bottom: 100 })}>设置距离底部 100px</Button>
    <Button onClick={() => notification.config({ className: 'custom-name' })}>更改 classname</Button>
    <br />
    <br />
    <Button
      onClick={() =>
        notification.config({
          placement: undefined,
          top: undefined,
          bottom: undefined,
          duration: undefined,
        })
      }
    >
      重设
    </Button>
    <Button
      type="primary"
      onClick={() =>
        notification.success({
          message: '成功',
          description: overlengthStr,
        })
      }
    >
      点我打开
    </Button>

    <Divider text="更新" />
    <Button
      type="primary"
      onClick={() => {
        notification.warning({
          key: 'need_update',
          message: 'Ready to update',
          duration: 5000,
          description: 'Will update after 2 seconds...',
        });
        setTimeout(() => {
          notification.success({
            key: 'need_update',
            message: 'Success',
            description: 'Update success!',
          });
        }, 2000);
      }}
    >
      点我打开
    </Button>
  </div>
);

const title = 'Ready to updateReady, Ready to updateReady,';

export const Overflow = () => (
  <Space>
    <Button
      onClick={() => {
        notification.warning({
          key: 'need_update',
          message: title + title + title,
          duration: null,
          description: 'Will close after 10 seconds...',
        });
      }}
    >
      标题超长文案
    </Button>
  </Space>
);
