// @universe-design/react/src/notification/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Basic from '../demo/0.basic';
import C1Type from '../demo/1.type';
import C2Position from '../demo/2.position';
import C3Btn from '../demo/3.btn';
import C4Update from '../demo/4.update';
import C5Style from '../demo/5.style';
import C6Icon from '../demo/6.icon';
import C7Close from '../demo/7.close';
import C8Config from '../demo/8.config';
if (!(C0Basic as Story).storyName) {
  (C0Basic as Story).storyName = '0. 基础用法';
  (C0Basic as Story).parameters = {
    docs: {
      description: {
        story: '简单的用法。3 秒后自动关闭。\n\n',
      },
    },
  };
}
if (!(C1Type as Story).storyName) {
  (C1Type as Story).storyName = '1. 通知的类型';
  (C1Type as Story).parameters = {
    docs: {
      description: {
        story: '通知提醒框有 5 种不同的类型，分别为`info`、`success`、`warning`、`error`、`normal`。\n\n',
      },
    },
  };
}
if (!(C2Position as Story).storyName) {
  (C2Position as Story).storyName = '2. 通知的位置';
  (C2Position as Story).parameters = {
    docs: {
      description: {
        story: '通知提醒框有 4 种不同的弹出位置，分别为`topLeft`、`topRight`、`bottomLeft`、`bottomRight`。\n\n',
      },
    },
  };
}
if (!(C3Btn as Story).storyName) {
  (C3Btn as Story).storyName = '3. 自定义操作按钮';
  (C3Btn as Story).parameters = {
    docs: {
      description: {
        story: '通过设置`btn`可以添加操作按钮。\n\n',
      },
    },
  };
}
if (!(C4Update as Story).storyName) {
  (C4Update as Story).storyName = '4. 动态更新通知设置';
  (C4Update as Story).parameters = {
    docs: {
      description: {
        story: '通过指定`key`，可以更新已经存在的通知提醒框。\n\n',
      },
    },
  };
}
if (!(C5Style as Story).storyName) {
  (C5Style as Story).storyName = '5. 自定义样式';
  (C5Style as Story).parameters = {
    docs: {
      description: {
        story: '通过设置`style`和`className`，为通知提醒框添加自定义样式。\n\n',
      },
    },
  };
}
if (!(C6Icon as Story).storyName) {
  (C6Icon as Story).storyName = '6. 自定义图标';
  (C6Icon as Story).parameters = {
    docs: {
      description: {
        story: '通过设置`icon`可以自定义图标。\n\n',
      },
    },
  };
}
if (!(C7Close as Story).storyName) {
  (C7Close as Story).storyName = '7. 手动关闭提醒';
  (C7Close as Story).parameters = {
    docs: {
      description: {
        story:
          '设置`duration`为`null`或`0`时，通知提醒框不会自动关闭。设置`key`后，可以通过`notification.close(key)`关闭指定`key`的通知提醒框。使用`notification.destroy()`, 可以关闭所有已经打开的通知提醒框。\n\n',
      },
    },
  };
}
if (!(C8Config as Story).storyName) {
  (C8Config as Story).storyName = '8. 设置全局生效的配置';
  (C8Config as Story).parameters = {
    docs: {
      description: {
        story:
          '通过`notification.config(options)`为通知提醒框设置全局生效的配置。通知提醒框会优先使用`Method`中的相关配置。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Notification/官网示例',
  tags: ['autodocs'],
};

export { C0Basic, C1Type, C2Position, C3Btn, C4Update, C5Style, C6Icon, C7Close, C8Config };
