// @universe-design/react/es/date-picker/panels/date-body/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React from 'react';
import moment from 'moment';
import clsx from 'clsx';
import PanelBody from '../base/Body';
import { useCssScope, useLocaleReceiver } from '../../../config-provider/hooks';
import Cell, { useDateCellInteractiveAreaCls } from '../../cell';
import { BodyConfig } from '../base/config';
import { getWeekStartDate } from '../../utils/date';
import { useDateCalendarUIContext } from '../DateCalendar/DateCalendarUIContext';
import { isSelectCell } from '../DateCalendar/utils';
import WeekNumberWrapper from '../../weekNumberWrapper';
function DateBody() {
  const {
    selectedDates = [],
    selectedRange,
    pickerValue,
    previewDate,
    cellRenderer,
    startOfWeek,
    noOutOfViewWeek,
    shouldCellDisabled,
    hideWeekNumbers = true
  } = useDateCalendarUIContext();
  const prefixCls = useCssScope('picker')();
  const locale = useLocaleReceiver('datePicker');
  const datePanelPrefixCls = `${prefixCls}-date-panel`;
  const dateCellInteractiveAreaCls = useDateCellInteractiveAreaCls();
  const baseDate = getWeekStartDate(pickerValue, startOfWeek);
  const DAYLABELS = locale.weekdaysMin;
  const dayLabels = [...DAYLABELS, ...DAYLABELS].slice(startOfWeek, startOfWeek + 7);
  let {
    rowNum
  } = BodyConfig.date;
  if (noOutOfViewWeek) {
    const {
      colNum
    } = BodyConfig.date;
    // 展示的天数 = 当前月天数 + 展示的前一个月天数
    const shouldShowDays = pickerValue.daysInMonth() + (pickerValue.clone().startOf('month').day() - startOfWeek + 7) % 7;
    const redundantLines = Math.floor(Math.abs(rowNum * colNum - shouldShowDays) / colNum);
    rowNum -= redundantLines;
  }
  return /*#__PURE__*/React.createElement(WeekNumberWrapper, {
    hideWeekNumbers: hideWeekNumbers,
    rowNum: rowNum,
    colNum: BodyConfig.date.colNum,
    baseDate: baseDate
  }, /*#__PURE__*/React.createElement(PanelBody, _extends({}, BodyConfig.date, {
    rowNum: rowNum,
    autoHeight: noOutOfViewWeek,
    cellClassName: `${datePanelPrefixCls}-cell`,
    baseDate: baseDate,
    getCellStatus: date => {
      const now = moment();
      // const [startDate = null, endDate = null] = rangedValue;
      const isOutOfView = !pickerValue.isSame(date, 'month');
      const isToday = date.isSame(now, 'day');
      const isRangePreview = previewDate ? previewDate(date) : false;
      const [startDate, endDate] = selectedRange || [];
      return {
        isToday,
        isDisabled: shouldCellDisabled(date, 'date'),
        isOutOfView,
        isRange: Boolean(startDate && endDate && date.isSameOrAfter(startDate, 'day') && date.isSameOrBefore(endDate, 'day')),
        isSelected: isSelectCell({
          selectedDates,
          selectedRange,
          mode: 'date',
          date
        }),
        isRangePreview
      };
    },
    getCellDate: (d, offset) => d.clone().add(offset, 'day'),
    dateRender: d => {
      const node = /*#__PURE__*/React.createElement("div", {
        className: dateCellInteractiveAreaCls
      }, d.format('D'));
      if (typeof cellRenderer === 'function') {
        return cellRenderer(d, {
          viewMode: 'date',
          defaultNode: node
        });
      }
      return node;
    }
  }), /*#__PURE__*/React.createElement("div", {
    className: clsx(`${datePanelPrefixCls}-weeks`)
  }, dayLabels.map((label, index) => /*#__PURE__*/React.createElement(Cell, {
    key: label + index,
    className: clsx(`${datePanelPrefixCls}-cell`, `${datePanelPrefixCls}-week-cell`),
    mode: "text"
  }, label)))));
}
export default DateBody;
