// @byted-service/env/lib/internal/net/ip.js
"use strict";
/**
 * 以下来自 go 标准库 -> net/ip.go
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseCIDR = exports.parseIP = exports.parseIPv6 = exports.parseIPv4 = exports.Mask = exports.CIDRMask = exports.IPv4 = exports.v4InV6Prefix = exports.IPv6len = exports.IPv4len = void 0;
const parse_1 = require("./parse");
exports.IPv4len = 4;
exports.IPv6len = 16;
exports.v4InV6Prefix = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0xff, 0xff];
/**
 * IPv4 returns the IP address (in 16-byte form) of the
 * IPv4 address a.b.c.d.
 */
const IPv4 = (a, b, c, d) => {
    const p = exports.v4InV6Prefix.slice(0);
    p[12] = a;
    p[13] = b;
    p[14] = c;
    p[15] = d;
    return p;
};
exports.IPv4 = IPv4;
/**
 * CIDRMask returns an IPMask consisting of 'ones' 1 bits
 * followed by 0s up to a total length of 'bits' bits.
 * For a mask of this form, CIDRMask is the inverse of IPMask.Size.
 */
const CIDRMask = (ones, bits) => {
    // if (bits !== 8 * IPv4len && bits !== 8 * IPv6len) {
    //     return null;
    // }
    // if (ones < 0 || ones > bits) {
    //     return null;
    // }
    const l = bits / 8;
    const m = [];
    let n = ones;
    for (let i = 0; i < l; i++) {
        if (n >= 8) {
            m[i] = 0xff;
            n -= 8;
            continue;
        }
        m[i] = (0xff >> n) ^ 0xff;
        n = 0;
    }
    return m;
};
exports.CIDRMask = CIDRMask;
const allFF = (b) => {
    for (let i = 0; i < b.length; i++) {
        const c = b[i];
        if (c !== 0xff) {
            return false;
        }
    }
    return true;
};
/**
 * Mask returns the result of masking the IP address ip with mask.
 */
const Mask = (ip, mask) => {
    if (mask.length === exports.IPv6len &&
        ip.length === exports.IPv4len &&
        allFF(mask.slice(0, 12))) {
        mask = mask.slice(12);
    }
    if (mask.length === exports.IPv4len &&
        ip.length === exports.IPv6len &&
        ip.slice(0, 12).join('.') === exports.v4InV6Prefix.join('.')) {
        ip = ip.slice(12);
    }
    const n = ip.length;
    if (n !== mask.length) {
        return null;
    }
    const out = [];
    for (let i = 0; i < n; i++) {
        out[i] = ip[i] & mask[i];
    }
    return out;
};
exports.Mask = Mask;
/**
 * Parse IPv4 address (d.d.d.d).
 */
const parseIPv4 = (s) => {
    const p = new Array(exports.IPv4len).fill(0);
    for (let i = 0; i < exports.IPv4len; i++) {
        if (s.length === 0) {
            // Missing octets.
            return null;
        }
        if (i > 0) {
            if (s[0] !== '.') {
                return null;
            }
            s = s.slice(1);
        }
        const [n, c, ok] = (0, parse_1.dtoi)(s);
        if (!ok || n > 0xff) {
            return null;
        }
        s = s.slice(c);
        p[i] = n & 0xff;
    }
    if (s.length !== 0) {
        return null;
    }
    return (0, exports.IPv4)(p[0], p[1], p[2], p[3]);
};
exports.parseIPv4 = parseIPv4;
/**
 * parseIPv6 parses s as a literal IPv6 address described in RFC 4291
 * and RFC 5952.
 */
const parseIPv6 = (ss) => {
    let s = ss;
    const ip = new Array(exports.IPv6len).fill(0);
    let ellipsis = -1; // position of ellipsis in ip
    // Might have leading ellipsis
    if (s.length >= 2 && s[0] === ':' && s[1] === ':') {
        ellipsis = 0;
        s = s.slice(2);
        // Might be only ellipsis
        if (s.length === 0) {
            return ip;
        }
    }
    // Loop, parsing hex numbers followed by colon.
    let i = 0;
    for (; i < exports.IPv6len;) {
        // Hex number.
        const [n, c, ok] = (0, parse_1.xtoi)(s);
        if (!ok || n > 0xffff) {
            return null;
        }
        // If followed by dot, might be in trailing IPv4.
        if (c < s.length && s[c] === '.') {
            if (ellipsis < 0 && i !== exports.IPv6len - exports.IPv4len) {
                // Not the right place.
                return null;
            }
            /* istanbul ignore next */
            if (i + exports.IPv4len > exports.IPv6len) {
                // Not enough room.
                return null;
            }
            const ip4 = (0, exports.parseIPv4)(s);
            if (ip4 === null) {
                return null;
            }
            ip[i] = ip4[12];
            ip[i + 1] = ip4[13];
            ip[i + 2] = ip4[14];
            ip[i + 3] = ip4[15];
            s = '';
            i += exports.IPv4len;
            break;
        }
        // Save this 16-bit chunk.
        ip[i] = (n >> 8) & 0xff;
        ip[i + 1] = n & 0xff;
        i += 2;
        // Stop at end of string.
        s = s.slice(c);
        if (s.length === 0) {
            break;
        }
        // Otherwise must be followed by colon and more.
        if (s[0] !== ':' || s.length === 1) {
            return null;
        }
        s = s.slice(1);
        // Look for ellipsis.
        if (s[0] === ':') {
            /* istanbul ignore next */
            if (ellipsis >= 0) {
                // already have one
                return null;
            }
            ellipsis = i;
            s = s.slice(1);
            if (s.length === 0) {
                // can be at end
                break;
            }
        }
    }
    // Must have used entire string.
    if (s.length !== 0) {
        return null;
    }
    // If didn't parse enough, expand ellipsis.
    if (i < exports.IPv6len) {
        if (ellipsis < 0) {
            return null;
        }
        const n = exports.IPv6len - i;
        for (let j = i - 1; j >= ellipsis; j--) {
            ip[j + n] = ip[j];
        }
        for (let j = ellipsis + n - 1; j >= ellipsis; j--) {
            ip[j] = 0;
        }
    }
    else if (ellipsis >= 0) {
        // Ellipsis must represent at least one 0 group.
        return null;
    }
    return ip;
};
exports.parseIPv6 = parseIPv6;
// ParseIP parses s as an IP address, returning the result.
// The string s can be in dotted decimal ("192.0.2.1")
// or IPv6 ("2001:db8::68") form.
// If s is not a valid textual representation of an IP address,
// ParseIP returns nil.
const parseIP = (s) => {
    for (let i = 0; i < s.length; i++) {
        switch (s[i]) {
            case '.':
                return (0, exports.parseIPv4)(s);
            case ':':
                return (0, exports.parseIPv6)(s);
        }
    }
    return null;
};
exports.parseIP = parseIP;
/**
 * ParseCIDR parses s as a CIDR notation IP address and prefix length,
 * like "192.0.2.0/24" or "2001:db8::/32", as defined in
 * RFC 4632 and RFC 4291.
 *
 * It returns the IP address and the network implied by the IP and
 * prefix length.
 * For example, ParseCIDR("192.0.2.1/24") returns the IP address
 * 192.0.2.1 and the network 192.0.2.0/24.
 */
const parseCIDR = (s) => {
    const i = s.indexOf('/');
    if (i < 0) {
        return [, , new Error('invalid CIDR address: ' + s)];
    }
    const addr = s.slice(0, i);
    const mask = s.slice(i + 1);
    let iplen = exports.IPv4len;
    let ip = (0, exports.parseIPv4)(addr);
    if (!ip) {
        iplen = exports.IPv6len;
        ip = (0, exports.parseIPv6)(addr);
    }
    const [n, index, ok] = (0, parse_1.dtoi)(mask);
    if (!ip || !ok || index !== mask.length || n < 0 || n > 8 * iplen) {
        return [, , new Error('invalid CIDR address: ' + s)];
    }
    const m = (0, exports.CIDRMask)(n, 8 * iplen);
    return [ip, { IP: (0, exports.Mask)(ip, m), Mask: m }, null];
};
exports.parseCIDR = parseCIDR;
