// @byted-service/fetch-core/lib/fetch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fetch = exports.protocols = void 0;
const logger_1 = require("@byted-service/logger");
const fetch_context_1 = require("./fetch_context");
const logger_2 = require("./logger");
const metrics_1 = require("./metrics");
const span_1 = require("./span");
const consts_1 = require("./consts");
/* istanbul ignore next */
exports.protocols = new Map([
    ['http', (ctx) => ctx.fetch()],
    ['https', (ctx) => ctx.fetch()],
]);
const logger = new logger_1.Logger();
async function fetch(url, options = {}) {
    const ctx = new fetch_context_1.FetchContext(url, options);
    const enableLogger = typeof options.enableTraceLogger === 'boolean'
        ? options.enableTraceLogger
        : typeof options.enableLogger === 'boolean'
            ? options.enableLogger
            : typeof options.disableLogger === 'boolean'
                ? !options.disableLogger
                : false;
    const enableMetrics = typeof options.enableTraceMetrics === 'boolean'
        ? options.enableTraceMetrics
        : typeof options.enableMetrics === 'boolean'
            ? options.enableMetrics
            : typeof options.disableMetrics === 'boolean'
                ? !options.disableMetrics
                : false;
    const enableBytedtrace = typeof options.enableBytedtrace === 'boolean'
        ? options.enableBytedtrace
        : typeof options.disableTracer === 'boolean'
            ? !options.disableTracer
            : true;
    let requestHandle = exports.protocols.get(ctx.remoteProtocol);
    if (!requestHandle) {
        logger.log({
            logId: ctx.logId,
            level: 'warn',
            message: `unknow protocol: ${ctx.remoteProtocol}, use fetch as default handle`,
        });
        requestHandle = () => ctx.fetch();
    }
    if (enableBytedtrace) {
        (0, span_1.startSpan)(ctx);
    }
    // 返回的不是一个 response 的情况，暂不考虑
    let response = null;
    try {
        let finalError = null;
        const retry = ctx.retry + 1;
        while (ctx.times < retry) {
            if (ctx.times && options.retryDelayFunction) {
                await options.retryDelayFunction({
                    currentRetryTimes: ctx.times,
                    targetRetryTimes: ctx.retry,
                });
            }
            try {
                response = await requestHandle(ctx);
                break;
            }
            catch (err) {
                finalError = err;
                ctx.times++;
                ctx.resetTimer();
            }
        }
        if (ctx.times === retry) {
            throw finalError;
        }
        if (!response) {
            throw new Error(`response must exists, there is some bug in protocol: ${ctx.remoteProtocol}`);
        }
        ctx.status = response.status;
        ctx.statusText = encodeURIComponent(response.statusText);
        response.closeSpan = consts_1.asyncNoop;
        if (enableBytedtrace) {
            response.closeSpan = (0, span_1.createCloseSpan)(ctx, response);
        }
        return response;
    }
    catch (ex) {
        ctx.status = 500;
        ctx.statusText = encodeURIComponent(String(ex));
        if (enableBytedtrace) {
            (0, span_1.closeSpan)(ctx, null, ex, {});
        }
        throw ex;
    }
    finally {
        ctx.timer.stop();
        if (enableMetrics) {
            (0, metrics_1.sendMetrics)(ctx);
        }
        if (enableLogger) {
            (0, logger_2.sendLogger)(ctx);
        }
        if (!(response && enableBytedtrace)) {
            ctx.release();
        }
    }
}
exports.fetch = fetch;
let defaultTimeout = 3000;
fetch.setDefaultTimeout = (timeout) => (defaultTimeout = timeout);
fetch.getDefaultTimeout = () => defaultTimeout;
fetch.register = (protocol, handle) => {
    if (typeof protocol !== 'string') {
        throw new Error('protocol must be string');
    }
    if (typeof handle !== 'function') {
        throw new Error('handle must be function');
    }
    exports.protocols.set(protocol, handle);
};
