// @byted-service/fetch-core/lib/fetch_context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FetchContext = void 0;
const tslib_1 = require("tslib");
const url_1 = require("url");
const node_fetch_1 = tslib_1.__importStar(require("node-fetch"));
const env_1 = require("@byted-service/env");
const logid_1 = tslib_1.__importDefault(require("@byted-service/logid"));
const logger_1 = require("@byted-service/logger");
const MetaInfo = tslib_1.__importStar(require("@byted-service/metainfo"));
const timer_1 = require("@byted-service/timer");
const fetch_1 = require("./fetch");
const helper_1 = require("./helper");
const consts_1 = require("./consts");
const net_1 = require("net");
class FetchContext {
    /** @deprecated use ctx.toMethod instead */
    get handleMethod() {
        return this.toMethod;
    }
    /** @deprecated use ctx.loggerTags instead */
    get customLoggerTags() {
        return this.loggerTags;
    }
    /** @deprecated use ctx.metricsTags instead */
    get customMetricsTags() {
        return this.metricsTags;
    }
    constructor(url, options) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j;
        this.options = options;
        this.status = 0;
        this.statusText = '';
        this.times = 0;
        if (typeof url === 'string') {
            url = new url_1.URL(url);
        }
        const protocol = url.protocol.slice(0, -1);
        const originalHost = url.host;
        // 自定义协议需要降级为 http 处理，剩下的部分由 handle 内部执行决定
        if (!url.origin || url.origin === 'null') {
            // 规避 PSM 规范与 W3C DNS Name Spec 不完全一致问题，增加后缀，非 http Protocol 实现中使用 ctx.to 绕过
            if (!(0, net_1.isIP)(url.hostname) && /\.\d+$/.test(url.hostname)) {
                url.hostname += consts_1.hostnameSuffix;
            }
            url = new url_1.URL(`http${url.toString().slice(protocol.length)}`);
        }
        // <!-- whatwg/fetch standard -->
        this.url = url;
        this.method = (options.method || 'GET').toUpperCase();
        this.body = options.body;
        this.headers = new node_fetch_1.Headers(options.headers || {});
        this.redirect = options.redirect;
        this.signal = options.signal;
        // <!-- @byted-service/fetch-core -->
        this.fromMethod =
            options.fromMethod ||
                ((_a = options.ctx) === null || _a === void 0 ? void 0 : _a.handleMethod) ||
                'byted_service_fetch_unknown_from_method';
        this.toMethod =
            options.toMethod ||
                options.handleMethod ||
                ((_b = options.ctx) === null || _b === void 0 ? void 0 : _b.handleMethod) ||
                'byted_service_fetch_unknown_to_method';
        this.logId = options.logId || ((_c = options.ctx) === null || _c === void 0 ? void 0 : _c.logId) || (0, logid_1.default)();
        this.stressTag = options.stressTag || ((_d = options.ctx) === null || _d === void 0 ? void 0 : _d.stressTag) || '';
        this.env = options.env || ((_e = options.ctx) === null || _e === void 0 ? void 0 : _e.env) || '';
        this.from = options.from || (0, env_1.getPSM)();
        this.fromCluster = options.fromCluster || (0, env_1.getCluster)();
        this.fromIdc = options.fromIdc || (0, env_1.getIDC)();
        this.to = options.to || options.psm || originalHost;
        this.toCluster =
            options.toCluster ||
                options.cluster ||
                ((_f = options.ctx) === null || _f === void 0 ? void 0 : _f.toCluster) ||
                'default';
        this.toIdc = options.toIdc || options.idc || ((_g = options.ctx) === null || _g === void 0 ? void 0 : _g.toIdc) || '';
        this.logger =
            options.traceLogger ||
                options.logger ||
                (options.loggerOptions
                    ? new logger_1.Logger(options.loggerOptions)
                    : consts_1.defaultLogger);
        this.loggerTags = options.traceLoggerTags || options.loggerTags || {};
        this.location = options.location || (0, logger_1.getStackFilePath)(3);
        this.metrics =
            options.traceMetrics || options.metrics || consts_1.defaultMetrics;
        this.metricsTags =
            options.traceMetricsTags || options.metricsTags || {};
        this.span = options.span || ((_h = options.ctx) === null || _h === void 0 ? void 0 : _h.span);
        this.trace = options.trace || ((_j = options.ctx) === null || _j === void 0 ? void 0 : _j.trace);
        this.retry = options.retry || 0;
        this.agent = (0, helper_1.proxyAgent)(url, options) || options.agent;
        this.timeout =
            typeof options.timeout !== 'number'
                ? fetch_1.fetch.getDefaultTimeout()
                : options.timeout;
        this.timer = (0, timer_1.createTimer)(this.timeout);
        this.remoteProtocol = protocol;
        this.encodeUnicodeHeader = options.encodeUnicodeHeader || false;
        MetaInfo.copyMetaInfo(this.options, this);
        if (options.ctx) {
            MetaInfo.copyMetaInfo(options.ctx, this);
        }
        if (options.customTraceErrorDecider) {
            this.customTraceErrorDecider = options.customTraceErrorDecider;
        }
    }
    /**
     * 由于 lerna monorepo 内广泛应用了软链接，但 ts 自身对于软链接的支持有限。所以这里没法使用 ts 的类型推导能力，只能手动订正类型
     * https://github.com/microsoft/TypeScript/issues/29808
     */
    getFetchOptions() {
        const timeout = this.timer.remain();
        if (this.timer.timeout !== 0 && timeout === 0) {
            throw new node_fetch_1.FetchError(`network timeout at: ${this.url.toString()}`, 'request-timeout');
        }
        if (!this.headers.get('x-tt-logid') && this.logId) {
            this.headers.set('x-tt-logid', this.logId);
        }
        if (!this.headers.get('x-tt-env') && this.env) {
            this.headers.set('x-tt-env', this.env);
        }
        if (!this.headers.get('x-tt-stress') && this.stressTag) {
            this.headers.set('x-tt-stress', this.stressTag);
        }
        if (!this.headers.get('x-tt-trace') && this.trace) {
            this.headers.set('x-tt-trace', this.trace);
        }
        /**
         * https://bytedance.feishu.cn/wiki/wikcnopD317Jwv7t3YHVqs2d6Xa
         */
        if (!this.headers.get('x-use-ppe') &&
            (0, env_1.isPPE)() &&
            !this.options.disableUsePPEHeader) {
            this.headers.set('x-use-ppe', '1');
        }
        if (MetaInfo.hasMetaInfo(this)) {
            const headers = {};
            const newCtx = {};
            MetaInfo.copyMetaInfo(this, newCtx);
            MetaInfo.transferForward(newCtx);
            MetaInfo.toHTTPHeader(newCtx, headers);
            const keys = Object.keys(headers);
            for (let i = 0; i < keys.length; i++) {
                let value = headers[keys[i]];
                if (this.encodeUnicodeHeader) {
                    value = (0, helper_1.encodeUnicodeText)(value);
                }
                this.headers.set(keys[i], value);
            }
        }
        return Object.assign(Object.assign({}, this.options), { method: this.method, body: this.body, headers: this.headers, redirect: this.redirect, signal: this.signal, agent: this.agent, timeout });
    }
    async fetch() {
        return (await (0, node_fetch_1.default)(this.url, this.getFetchOptions()));
    }
    release() {
        if (this.timer) {
            (0, timer_1.releaseTimer)(this.timer);
            this.timer = null;
        }
    }
    resetTimer() {
        this.timer = (0, timer_1.createTimer)(this.timeout);
    }
}
exports.FetchContext = FetchContext;
