// @byted-service/fetch-core/lib/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sendLogger = void 0;
const net_1 = require("net");
const helper_1 = require("./helper");
const sendLogger = (ctx) => {
    ctx.logger.log({
        logId: ctx.logId,
        level: ctx.status >= 200 && ctx.status < 400 ? 'trace' : 'error',
        location: ctx.location,
        tags: Object.assign({ 
            // basic
            method: ctx.toMethod, from_method: ctx.fromMethod, env: ctx.env || 'prod', stress_tag: ctx.stressTag || '-', 
            // from / to
            to: ctx.to, to_cluster: ctx.toCluster, to_idc: ctx.toIdc || '-', 
            // call
            status: ctx.status, status_message: ctx.statusText, cost: ctx.timer.cost('μs'), retry: ctx.times, 
            // ip
            mesh: ctx.mesh === '1' ? '1' : '0', ipv6: ctx.rip && (0, net_1.isIPv6)((0, helper_1.parseAddr)(ctx.rip).host) ? '1' : '0', rip: ctx.rip || '', 
            // fetch only
            remote_protocol: ctx.remoteProtocol }, ctx.loggerTags),
    });
};
exports.sendLogger = sendLogger;
