// @byted-service/fetch-core/lib/span.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.dumpResponse = exports.dumpRequest = exports.closeSpan = exports.createCloseSpan = exports.startSpan = exports.logger = void 0;
const url_1 = require("url");
const stream_1 = require("stream");
const util_1 = require("util");
const logger_1 = require("@byted-service/logger");
const bytedtrace_1 = require("@byted-service/bytedtrace");
const metrics_1 = require("./metrics");
/* istanbul ignore next */
const isAnyArrayBuffer = util_1.types.isAnyArrayBuffer || (() => false);
const spanClosedSym = Symbol.for('#spanClosed');
exports.logger = new logger_1.Logger();
const startSpan = (ctx) => {
    const tracer = bytedtrace_1.BytedTracer.defaultBytedTracer();
    let span = ctx.span;
    if (!span) {
        span = ctx.span = tracer.startClientSpan(ctx.toMethod, {
            logId: ctx.logId,
        });
    }
    if (span.isForwardPropagate()) {
        const headers = {};
        const tracer = bytedtrace_1.BytedTracer.defaultBytedTracer();
        tracer.bytedInject(span.getContext(), bytedtrace_1.PropagatorTypes.HTTP_HEADER, headers);
        const keys = Object.keys(headers);
        for (let i = 0; i < keys.length; i++) {
            ctx.headers.set(keys[i], headers[keys[i]]);
        }
    }
};
exports.startSpan = startSpan;
const createCloseSpan = (ctx, response) => {
    const fn = async (options = {}) => {
        const span = ctx.span;
        if (span[spanClosedSym]) {
            return;
        }
        span[spanClosedSym] = true;
        try {
            await (0, exports.closeSpan)(ctx, response, null, options);
        }
        catch (err) {
            exports.logger.warn('Failed to close span "%s". %s', span.spanId, err);
        }
        finally {
            ctx.release();
        }
    };
    process.nextTick(fn);
    return fn;
};
exports.createCloseSpan = createCloseSpan;
const closeSpan = async (ctx, resp, error, options) => {
    const tracer = bytedtrace_1.BytedTracer.defaultBytedTracer();
    const span = ctx.span;
    const errorCode = error ? -1 : ctx.status;
    if (!span.isSampled()) {
        if (resp) {
            const headers = {};
            for (const key of resp.headers.keys()) {
                headers[key] = resp.headers.get(key) || '';
            }
            const context = tracer.bytedExtract(bytedtrace_1.PropagatorTypes.HTTP_HEADER, headers);
            /* istanbul ignore next */
            if (context.isValid()) {
                span.setPostTrace(context);
            }
        }
        if (error) {
            span.setPostTrace();
        }
    }
    span.setServiceType('http');
    span.setComponent('byted_service_fetch');
    // basic
    span.setToService(ctx.to);
    span.setToMethod(ctx.toMethod);
    span.setTag('env', ctx.env || 'prod');
    span.setTag('stress_tag', ctx.stressTag || '-');
    // from / to
    span.setTag('to', ctx.to);
    span.setTag('to_cluster', ctx.toCluster);
    span.setTag('to_idc', ctx.toIdc || '-');
    span.setToCluster(ctx.toCluster);
    if (ctx.toIdc) {
        span.setToDc(ctx.toIdc);
        // span.setTag('to_idc', ctx.toIdc);
    }
    // call
    span.setTag('retry', ctx.retry);
    span.setTag('http.method', ctx.method);
    span.setTag('http.url', ctx.url.pathname);
    span.setTag('http.host', ctx.url.host);
    span.setTag('http.status_code', ctx.status);
    span.setTag('http.param', ctx.url.searchParams.toString());
    span.setStatusCode(errorCode);
    const _reportTraceError = (isError) => {
        span.setIsError(isError);
        span.setErrorMessage(ctx.statusText);
        if (error) {
            span.addEvent(bytedtrace_1.LogEvent.createErrorLogEvent((0, util_1.format)('err-%d', errorCode)).setContent((0, util_1.format)('%s', error)));
        }
    };
    if (errorCode === -1) {
        _reportTraceError(true);
    }
    else if (ctx.customTraceErrorDecider &&
        typeof ctx.customTraceErrorDecider === 'function') {
        const isError = (await ctx.customTraceErrorDecider(errorCode)) === true;
        _reportTraceError(isError);
    }
    else if (errorCode >= 400) {
        _reportTraceError(true);
    }
    // ip
    span.setTag('mesh', ctx.mesh === '1' ? '1' : '0');
    span.setToAddr(ctx.rip || '-');
    // hertz
    span.setTag('to_env', ctx.env || 'prod');
    // fetch only
    span.setTag('remote_protocol', ctx.remoteProtocol);
    // http only
    if (options.request) {
        const request = (0, exports.dumpRequest)(ctx);
        span.setSendSize(request.length);
        span.addEvent(bytedtrace_1.LogEvent.createInfoLogEvent('http.request').setContent(request));
        const map = {};
        const params = ctx.url.searchParams;
        for (const key of params.keys()) {
            if (map[key] === undefined) {
                span.setTag(`http.param.${key}`, (map[key] = params.get(key) || ''));
            }
        }
    }
    if (options.response && resp) {
        const response = await (0, exports.dumpResponse)(resp);
        span.setRecvSize(response.length);
        span.addEvent(bytedtrace_1.LogEvent.createInfoLogEvent('http.response').setContent(response));
    }
    span.setFinishTime(Date.now() * 1000);
    span.finish();
    if (tracer.enableThriftStyleMetrics('nodejs.http.client')) {
        (0, metrics_1.sendMetrics)(ctx);
    }
};
exports.closeSpan = closeSpan;
const dumpRequestLine = (ctx) => {
    return `${ctx.method.toUpperCase()} ${ctx.url.pathname}${ctx.url.search} HTTP/1.1\n`;
};
const dumpHeaders = (headers) => {
    let headerStr = '';
    for (const key of headers.keys()) {
        headerStr += `${key}: ${headers.get(key) || ''}\n`;
    }
    return headerStr;
};
const dumpRequestBody = (body) => {
    if (!body) {
        // null | undefiend
        return '';
    }
    else if (typeof body === 'string') {
        // string
        return body;
    }
    else if (body instanceof url_1.URLSearchParams) {
        // URLSearchParams
        return body.toString();
        // } else if (isBlob()) {
    }
    else if (Buffer.isBuffer(body)) {
        // Buffer
        return body.toString();
    }
    else if (isAnyArrayBuffer(body) || body instanceof ArrayBuffer) {
        // ArrayBuffer
        /* istanbul ignore next */
        return Buffer.from(body).toString();
    }
    else if (ArrayBuffer.isView(body)) {
        // ArrayBufferView
        /* istanbul ignore next */
        return Buffer.from(body.buffer, body.byteOffset, body.byteLength).toString();
    }
    else if (body instanceof stream_1.Stream) {
        // Stream
        return '[stream Stream]';
        // } else if (isFormData()) {
    }
    else if (typeof body.toString === 'function') {
        return body.toString();
    }
    else {
        /* istanbul ignore next */
        return '[unknown Unknown]';
    }
};
const dumpRequest = (ctx) => {
    if (!ctx.headers.get('host')) {
        ctx.headers.set('host', ctx.url.host);
    }
    return (dumpRequestLine(ctx) +
        dumpHeaders(ctx.headers) +
        '\n' +
        dumpRequestBody(ctx.body));
};
exports.dumpRequest = dumpRequest;
const dumpStatusLine = (resp) => {
    return `HTTP/1.1 ${resp.status} ${resp.statusText}\n`;
};
const dumpResponseBody = async (resp) => {
    return await resp.clone().text();
};
const dumpResponse = async (resp, withBody) => {
    return (dumpStatusLine(resp) +
        dumpHeaders(resp.headers) +
        '\n' +
        (await dumpResponseBody(resp)));
};
exports.dumpResponse = dumpResponse;
