// @universe-design/react/es/select/context/useSelection.js
import identity from 'lodash/identity';
import React, { useCallback } from 'react';
import { useCssScope, useLocale } from '../../config-provider/hooks';
import { useDebounce, useEventCallback } from '../../hooks';
import { isFunction } from '../../util';
import { defaultTagRenderer } from '../SelectorTag';
import { SelectActionType, SelectSearchAction } from '../internal.interface';
import { useSelectFocus, useSelectInput, useSelectInputValue, useSelectOverlayVisible } from './internal';
import { useSelectDataMap } from './useSelectDataMap';
function selectReducer(option, multiple, prevState) {
  if (!option) {
    return {
      nextState: prevState
    };
  }
  if (prevState.has(option.value)) {
    return {
      nextState: prevState
    };
  }
  if (multiple) {
    const nextState = new Set(prevState);
    nextState.add(option.value);
    return {
      nextState
    };
  }
  return {
    nextState: new Set([option.value])
  };
}
function deselectReducer(key, prevState) {
  const nextState = new Set(prevState);
  nextState.delete(key);
  return {
    nextState
  };
}

/**
 * 判断一个 option 是否为 disable 状态
 * @param option
 * @param globalDisabled
 */
export function isDisabledOption(option, globalDisabled) {
  return globalDisabled || Boolean(option.disabled);
}

/**
 * getPermittedOptions
 * 从输入 key list 里筛选出可以操作的 option keys 列表
 */
function getPermittedOptionKeys(keys, globalDisabled, optionMap) {
  return keys.filter(key => {
    const opt = optionMap.get(key);
    if (!opt) {
      return !globalDisabled;
    }
    if (isDisabledOption(opt, globalDisabled)) {
      return false;
    }
    return true;
  });
}

// eslint-disable-next-line max-lines-per-function
export function useSelection(props) {
  var _ref;
  let config = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {
    dropdownCssScope: 'select',
    selectorCssScope: 'select'
  };
  const locale = useLocale();
  const {
    open,
    onDropdownVisibleChange,
    value: valueProp,
    defaultValue,
    onClear,
    onSelect,
    onDeselect,
    onSearch,
    onFocus,
    onBlur,
    searchValue,
    tokenSeparators,
    tabIndex,
    disabled = false,
    autoFocus = false,
    defaultOpen = false,
    autoClearSearchValue = true,
    options,
    onKeyDown,
    onKeyUp,
    size,
    tagRender: tagRenderProp,
    tagRenderer: tagRendererProp,
    id,
    notFoundContent = locale.global.notFoundContent,
    dropdownRender = identity,
    loading = false,
    dropdownMatchSelectWidth
  } = props;
  const multiple = props.mode === 'multiple';

  // multiple default showSearch
  const {
    showSearch = multiple,
    noOptionsContent = showSearch ? null : locale.global.noOptions
  } = props;
  const useSelectDataMapResult = useSelectDataMap({
    value: valueProp,
    defaultValue,
    options,
    valueInProps: 'value' in props
  });
  const {
    valueSet,
    setValueSet,
    optionMap,
    getOption
  } = useSelectDataMapResult;
  const {
    inputValue,
    setInputValue
  } = useSelectInputValue({
    searchValue,
    onSearch
  });
  const listRef = React.useRef(null);

  //  ========================== User Inputting ==========================

  const [inputting, setInputting] = React.useState(false);

  // ========================== Overlay Visible ==========================

  const {
    overlayVisible,
    setOverlayVisible
  } = useSelectOverlayVisible({
    disabled,
    defaultOpen,
    open,
    onDropdownVisibleChange,
    setInputting
  });

  // ========================== Focus / Blur ==========================

  const {
    focusState,
    inputRef,
    onContainerFocus,
    onContainerBlur,
    focus,
    blur,
    cancelDelayCloseOverlay
  } = useSelectFocus({
    autoFocus,
    onBlur,
    onFocus,
    disabled,
    overlayVisible,
    setOverlayVisible,
    setInputValue,
    setInputting
  });
  const [debouncedFocusState] = useDebounce(focusState, 0);
  const doClearInputValue = React.useCallback(() => {
    if (!multiple || autoClearSearchValue) {
      setInputValue('', {
        action: SelectSearchAction.AutoClear
      });
    }
  }, [autoClearSearchValue, setInputValue, multiple]);
  const handleSelectEvent = React.useCallback(action => {
    /**
     * 这里和 onChange 里都要调用 doClearInputValue
     * 1. 如果点击一个已选中的项目，触发 onSelect 但是不触发 onChange，这时候是需要 doClearInputValue 的
     * 2. 如果调用 set 函数，但是没有给 action 的话，只触发 onChange 不触发 onSelect，这时候也是需要 doClearInputValue 的
     */
    doClearInputValue();
    // onSelect 回调
    if (action.type === SelectActionType.Select) {
      isFunction(onSelect) && onSelect(action.payload, {
        option: getOption(action.payload)
      });
    } else if (action.type === SelectActionType.DeSelect) {
      isFunction(onDeselect) && onDeselect(action.payload, {
        option: getOption(action.payload)
      });
    }
  }, [doClearInputValue, getOption, onDeselect, onSelect]);
  const _onChange = useEventCallback((_options, action) => {
    doClearInputValue();
    if (!isFunction(props.onChange)) {
      return;
    }
    if (props.mode === 'multiple') {
      const meta = {
        option: _options.map(opt => getOption(opt)),
        action
      };
      props.onChange(_options, meta);
      return;
    }
    const option = _options.map(opt => getOption(opt))[0];
    props.onChange(option === null || option === void 0 ? void 0 : option.value, {
      option,
      action
    });
  });
  const toggle = React.useCallback(function (_value) {
    let option = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    if (!getPermittedOptionKeys([_value], disabled, optionMap).length) {
      return;
    }
    let action = {
      type: SelectActionType.Toggle,
      payload: undefined
    };
    setValueSet(s => {
      if ('closeOverlay' in option) {
        if (option.closeOverlay) {
          setOverlayVisible(false);
        }
      } else if (!multiple) {
        setOverlayVisible(false);
      }
      setTimeout(() => {
        var _inputRef$current;
        (_inputRef$current = inputRef.current) === null || _inputRef$current === void 0 ? void 0 : _inputRef$current.focus();
      });
      /**
       * Deselect
       * 1. single 不会 deselect
       * 2. multiple 如果存在 selectedMap 里，进行 deselect
       */
      if (valueSet.has(_value) && multiple) {
        action = {
          type: SelectActionType.DeSelect,
          payload: _value
        };
        handleSelectEvent(action);
        return deselectReducer(_value, s).nextState;
      }
      /**
       * Select
       * 1. option list 里面有的话进行选中
       */
      if (optionMap.get(_value)) {
        action = {
          type: SelectActionType.Select,
          payload: _value
        };
        handleSelectEvent(action);
        return selectReducer(optionMap.get(_value), multiple, s).nextState;
      }
      return s;
    }, nextState => {
      _onChange([...nextState.keys()], action);
    });
  }, [_onChange, disabled, handleSelectEvent, inputRef, multiple, optionMap, setOverlayVisible, setValueSet, valueSet]);
  const deselect = toggle;
  const set = React.useCallback(function (_value) {
    var _option$action;
    let option = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    if (!multiple && _value.length > 1) {
      throw new Error('Selection in single mode can not set _value which length gt 1');
    }
    if ('closeOverlay' in option) {
      if (option.closeOverlay) {
        setOverlayVisible(false);
      }
    } else if (!multiple) {
      setOverlayVisible(false);
    }
    setTimeout(() => {
      var _inputRef$current2;
      (_inputRef$current2 = inputRef.current) === null || _inputRef$current2 === void 0 ? void 0 : _inputRef$current2.focus();
    });

    // 单选不可以取消, 但需要触发 onSelect
    if (!multiple && ((_option$action = option.action) === null || _option$action === void 0 ? void 0 : _option$action.type) === SelectActionType.DeSelect) {
      // 触发 onSelect
      option.action && handleSelectEvent({
        ...option.action,
        type: SelectActionType.Select
      });
      return;
    }
    setValueSet(() => _value.reduce((pre, cur) => {
      const opt = getOption(cur);
      if (opt) {
        pre.add(cur);
      }
      return pre;
    }, new Set()), nextState => {
      var _option$action2;
      option.action && handleSelectEvent(option.action);
      _onChange([...nextState.keys()], (_option$action2 = option.action) !== null && _option$action2 !== void 0 ? _option$action2 : {
        type: SelectActionType.Set,
        payload: undefined
      });
    });
  }, [_onChange, getOption, handleSelectEvent, inputRef, multiple, setOverlayVisible, setValueSet]);
  const clear = React.useCallback(() => {
    setValueSet(() => new Set(), nextState => {
      setOverlayVisible(false);
      setTimeout(() => {
        var _inputRef$current3;
        (_inputRef$current3 = inputRef.current) === null || _inputRef$current3 === void 0 ? void 0 : _inputRef$current3.focus();
      });
      onClear === null || onClear === void 0 ? void 0 : onClear();
      _onChange([...nextState.keys()], {
        type: SelectActionType.Clear,
        payload: undefined
      });
    });
  }, [_onChange, inputRef, onClear, setOverlayVisible, setValueSet]);
  const deselectLast = React.useCallback(() => {
    const keys = [...valueSet.keys()];
    const keyToDeselect = keys[keys.length - 1];
    if (keyToDeselect) {
      deselect(keyToDeselect);
    }
  }, [deselect, valueSet]);
  const {
    onInputChange,
    onInputCompositionEnd,
    onInputCompositionStart,
    onInputPaste,
    onInternalKeyUp,
    onInternalKeyDown,
    compositionStatusRef
  } = useSelectInput({
    multiple,
    tokenSeparators,
    inputValue,
    setInputValue,
    set,
    deselectLast,
    overlayVisible,
    setOverlayVisible,
    onKeyDown,
    onKeyUp,
    listRef,
    setInputting,
    getOption
  });
  const cssScope = useCssScope(`${config.selectorCssScope}`);
  const dropdownCssScope = useCssScope(`${config.dropdownCssScope}__dropdown`);
  const selectorCssScope = useCssScope(`${config.selectorCssScope}__selector`);

  // ================ props.optionLabelProp ================
  let {
    optionLabelProp
  } = props;
  // 这里是有可能出现 children 的
  if (optionLabelProp === 'children' || optionLabelProp === undefined) {
    optionLabelProp = config.defaultOptionLabelProp;
  }
  const selectorLabelRenderer = useCallback(_value => {
    var _optionLabelProp;
    const option = getOption(_value);
    if (isFunction(optionLabelProp)) {
      return optionLabelProp(option);
    }
    const node = option[(_optionLabelProp = optionLabelProp) !== null && _optionLabelProp !== void 0 ? _optionLabelProp : 'label'];
    return node;
  }, [getOption, optionLabelProp]);
  const tagRenderer = (_ref = tagRenderProp !== null && tagRenderProp !== void 0 ? tagRenderProp : tagRendererProp) !== null && _ref !== void 0 ? _ref : defaultTagRenderer;
  return {
    focusState,
    disabled,
    blur,
    focus,
    inputRef,
    listRef,
    inputValue,
    overlayVisible,
    multiple,
    valueSet,
    getOption,
    set,
    select: toggle,
    deselect,
    toggle,
    clear,
    deselectLast,
    setInputValue,
    setOverlayVisible,
    selectorCssScope,
    dropdownCssScope,
    onContainerFocus,
    onContainerBlur,
    onInputChange,
    onInputCompositionEnd,
    onInputCompositionStart,
    onInputPaste,
    onInternalKeyDown,
    onInternalKeyUp,
    compositionStatusRef,
    showSearch,
    tabIndex,
    selectorLabelRenderer,
    cssScope,
    size,
    cancelDelayCloseOverlay,
    tagRenderer,
    id,
    // 若使用 focusState 来控制 inputting，在点击 option 时会短暂失焦，导致输入框中的内容闪动。因此改为 debounced 后的 focusState
    // inputting: showSearch && focusState && inputting,
    inputting: showSearch && debouncedFocusState && inputting,
    setInputting,
    noOptionsContent,
    notFoundContent,
    dropdownRender,
    loading,
    textOverflow: dropdownMatchSelectWidth === true
  };
}
