// @universe-design/react/src/notice/stories/homeless.stories.tsx
/** @jsxImportSource @emotion/react */
// also exported from '@storybook/react' if you can deal with breaking changes in 6.1
import React from 'react';
import type { Meta } from '@storybook/react';

import { Row, Col, Notice, Button, Space } from '../..';

const types = ['info', 'error', 'warning', 'success'] as const;
const titles = [
  {
    label: '短 title',
    message: '这是一条提示文本信息',
    description: '',
  },
  {
    label: '长信息',
    maxLine: 4,
    message: '一个简短的标题',
    description:
      '这是正文提示文本信息，文本信息过长建议增加标题，文本行数建议 4 行以内，过长情况下不易阅读，标题宽度与正文宽度等宽处理，过长情况下不易阅读，标题宽度与正文宽度等宽处理，过长情况下不易阅读，标题宽度与正文宽度等宽处理，过长情况下不易阅读，标题宽度与正文宽度等宽处理。',
  },
];
const closables = [
  {
    label: '可关闭',
    value: true,
  },
  {
    label: '不关闭',
    value: false,
  },
];
const aligns: {
  label: string;
  value: 'center' | 'left';
}[] = [
  {
    label: 'center',
    value: 'center',
  },
  {
    label: 'left',
    value: 'left',
  },
];

export default {
  title: 'Components/Notice/临时示例',
  component: Notice,
} satisfies Meta;

const actionNode = (
  <Space size="medium" style={{ margin: '-2px 0' }}>
    <Button type="text" color="primary" style={{ marginLeft: '-4px' }}>
      文字按钮
    </Button>
    <Button type="text" color="primary">
      文字按钮
    </Button>
  </Space>
);

export const AllNotices = () => (
  <div>
    <Row gutter={16}>
      {titles.map(title =>
        closables.map(closable =>
          aligns.map(align => {
            const key = `${title.label}, ${closable.label}, ${align.label}`;
            return (
              <Col key={key}>
                <h4>{key}</h4>
                <div css={{ width: 420, marginBottom: 16 }}>
                  <Notice
                    showIcon
                    closable={closable.value}
                    align={align.value}
                    message={title.message}
                    description={title.description}
                    maxLine={title.maxLine}
                  />
                </div>
              </Col>
            );
          }),
        ),
      )}
    </Row>
    <h4>多状态带描述</h4>
    <Notice
      showIcon
      key="123"
      type="info"
      message="这是正文提示文本信息"
      closable
      style={{ width: 640, marginBottom: 16 }}
      action={actionNode}
    />
    <Notice
      showIcon
      key="123"
      type="loading"
      message="这是正文提示文本信息"
      closable
      style={{ width: 640, marginBottom: 16 }}
      action={actionNode}
    />
    <Notice
      showIcon
      key="123"
      type="success"
      message="这是正文提示文本信息"
      closable
      style={{ width: 640, marginBottom: 16 }}
      action={actionNode}
    />
    <Notice
      showIcon
      key="123"
      type="warning"
      message="这是正文提示文本信息"
      closable
      style={{ width: 640, marginBottom: 16 }}
      action={actionNode}
    />
    <Notice
      showIcon
      key="123"
      type="error"
      message="这是正文提示文本信息"
      closable
      style={{ width: 640, marginBottom: 16 }}
      action={actionNode}
    />
    <Notice
      showIcon
      key="123"
      type="info"
      message="这是正文提示文本信息"
      closable
      style={{ width: 640, marginBottom: 16 }}
      action={actionNode}
    />
    <Notice
      showIcon
      key="123"
      type="info"
      message="一个大标题"
      description="这是正文提示文本信息"
      closable
      style={{ width: 640, marginBottom: 16 }}
      action={actionNode}
    />
    <Notice
      showIcon
      key="123"
      type="info"
      message="这是正文提示文本信息这是正文提示文本信息这是正文提示文本信息这是正文提示文本信息这是正文提示文本信息"
      closable
      style={{ width: 640, marginBottom: 16 }}
      action={actionNode}
    />
    {types.map(type => (
      <div key={type} css={{ width: 640, marginBottom: 16 }}>
        <Notice
          showIcon
          key={type}
          type={type}
          message="标题"
          closable
          description={
            <div>
              <div>
                这是正文提示文本信息，文本信息过长建议增加标题，文本行数建议 4
                行以内，过长情况下不易阅读，标题宽度与正文宽度等宽处理。
              </div>
            </div>
          }
          action={
            <Space size="medium" style={{ margin: '-2px 0' }}>
              <Button type="text" color="primary" style={{ marginLeft: '-4px' }}>
                文字按钮
              </Button>
              <Button type="text" color="primary">
                文字按钮
              </Button>
              <Button type="text" color="primary">
                文字按钮
              </Button>
            </Space>
          }
        />
      </div>
    ))}
  </div>
);
