// @universe-design/react/src/notice/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Basic from '../demo/0.basic';
import C1Style from '../demo/1.style';
import C12CustomIcon from '../demo/12.custom-icon';
import C2Align from '../demo/2.align';
import C2Closable from '../demo/2.closable';
import C3Description from '../demo/3.description';
import C4Icon from '../demo/4.icon';
import C6CustomAction from '../demo/6.custom-action';
import C7ControlledMode from '../demo/7.controlled-mode';
if (!(C0Basic as Story).storyName) {
  (C0Basic as Story).storyName = '0. 基础用法';
  (C0Basic as Story).parameters = {
    docs: {
      description: {
        story:
          '最简单的用法。适用于简短的警告提示。\n\n\n<style>\n.react-site-demo .ud__notice {\n  min-width: 150px;\n  max-width: 50%;\n}\n\n.react-site-demo .ud__notice:not(:last-child) {\n  margin-bottom: 24px;\n}\n</style>\n',
      },
    },
  };
}
if (!(C1Style as Story).storyName) {
  (C1Style as Story).storyName = '1. 四种样式';
  (C1Style as Story).parameters = {
    docs: {
      description: {
        story: '共有四种样式，`success`、`info`、`warning`、`error`。\n\n',
      },
    },
  };
}
if (!(C12CustomIcon as Story).storyName) {
  (C12CustomIcon as Story).storyName = '12. 自定义图标';
  (C12CustomIcon as Story).parameters = {
    docs: {
      description: {
        story: '可口的图标让信息类型更加醒目。\n\n',
      },
    },
  };
}
if (!(C2Align as Story).storyName) {
  (C2Align as Story).storyName = '2. 内容对齐方式';
  (C2Align as Story).parameters = {
    docs: {
      description: {
        story: '提供两种对齐方式，`left`、`center`。`center` 在同时有`message`和`description`属性时不生效。\n\n',
      },
    },
  };
}
if (!(C2Closable as Story).storyName) {
  (C2Closable as Story).storyName = '2. 可关闭的警告提示';
  (C2Closable as Story).parameters = {
    docs: {
      description: {
        story: '显示关闭按钮，点击可关闭警告提示。\n\n',
      },
    },
  };
}
if (!(C3Description as Story).storyName) {
  (C3Description as Story).storyName = '3. 含有辅助性文字介绍';
  (C3Description as Story).parameters = {
    docs: {
      description: {
        story: '含有辅助性文字介绍的警告提示。\n\n',
      },
    },
  };
}
if (!(C4Icon as Story).storyName) {
  (C4Icon as Story).storyName = '4. 图标';
  (C4Icon as Story).parameters = {
    docs: {
      description: {
        story: '可口的图标让信息类型更加醒目。\n\n',
      },
    },
  };
}
if (!(C6CustomAction as Story).storyName) {
  (C6CustomAction as Story).storyName = '5. 自定义操作';
  (C6CustomAction as Story).parameters = {
    docs: {
      description: {
        story: '使用文字按钮自定义操作。\n\n',
      },
    },
  };
}
if (!(C7ControlledMode as Story).storyName) {
  (C7ControlledMode as Story).storyName = '7. 受控模式';
  (C7ControlledMode as Story).parameters = {
    docs: {
      description: {
        story: '使用受控模式显示隐藏常驻提示。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Notice/官网示例',
  tags: ['autodocs'],
};

export {
  C0Basic,
  C1Style,
  C12CustomIcon,
  C2Align,
  C2Closable,
  C3Description,
  C4Icon,
  C6CustomAction,
  C7ControlledMode,
};
