// @universe-design/react/lib/tooltip/Tooltip.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.splitObject = exports.Tooltip = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _hooks = require("../config-provider/hooks");
var _usePopperSidecar = require("../config-provider/hooks/usePopperSidecar");
require("./style");
const splitObject = (obj, keys) => {
  const picked = {};
  const omitted = {
    ...obj
  };
  keys.forEach(key => {
    if (obj && key in obj) {
      picked[key] = obj[key];
      delete omitted[key];
    }
  });
  return {
    picked,
    omitted
  };
};

/**
 * @public
 * @docgen
 */
exports.splitObject = splitObject;
const Tooltip = exports.Tooltip = /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
  const {
    prefixCls,
    title,
    overlay,
    trigger = 'hover',
    // maxWidth, // max-width 现在加在了 popup 上可以直接用 overlayStyle 去覆盖
    placement = 'top',
    overlayStyle,
    overlayClassName,
    hideArrow: hideArrowProp = false,
    showArrow: showArrowProp = true,
    // adjustOverflow = true,
    children,
    onOverlayMouseEnter,
    onOverlayMouseLeave,
    mouseLeaveDelay = 0,
    wrapDisabledChildren = false,
    arrowHasStroke = false,
    ...restProps
  } = props;
  const cssScope = (0, _hooks.useCssScope)('tooltip', prefixCls);
  const Popper = (0, _usePopperSidecar.usePopperSidecar)();
  const showArrow = 'hideArrow' in props ? !hideArrowProp : showArrowProp;
  const getDisabledCompatibleChildren = _react.default.useCallback(element => {
    if (!element.props.disabled) {
      return element;
    }
    // Pick some layout related style properties up to span
    // Prevent layout bugs
    const {
      picked,
      omitted
    } = splitObject(element.props.style, ['position', 'left', 'right', 'top', 'bottom', 'float', 'display', 'zIndex']);
    const spanStyle = {
      display: 'inline-block',
      // default inline-block is important
      ...picked,
      cursor: 'not-allowed',
      width: element.props.block ? '100%' : null
    };
    const buttonStyle = {
      ...omitted,
      pointerEvents: 'none'
    };
    const child = /*#__PURE__*/(0, _react.cloneElement)(element, {
      style: buttonStyle,
      className: null
    });
    return /*#__PURE__*/_react.default.createElement("span", {
      style: spanStyle,
      className: (0, _clsx.default)(element.props.className, `${cssScope()}-disabled-compatible-wrapper`)
    }, child);
  }, [cssScope]);
  const mergedChildren = (0, _react.useMemo)(() => {
    if (! /*#__PURE__*/(0, _react.isValidElement)(children)) {
      return /*#__PURE__*/_react.default.createElement("span", {
        className: cssScope('__invalid-children-wrapper')
      }, children);
    }

    /**
     * if children is disabled Button or native disabled button,
     * wrap children with <span>,
     * because disabled button won't fire mouseenter/mouseleave/mouseover events for some browsers
     */
    if (wrapDisabledChildren) {
      return getDisabledCompatibleChildren(children);
    }
    return children;
  }, [children, cssScope, getDisabledCompatibleChildren, wrapDisabledChildren]);
  const popup = _react.default.useMemo(() => /*#__PURE__*/_react.default.createElement("div", {
    onMouseEnter: onOverlayMouseEnter,
    onMouseLeave: onOverlayMouseLeave,
    className: cssScope('-content'),
    role: "tooltip"
  }, title !== null && title !== void 0 ? title : overlay), [cssScope, onOverlayMouseEnter, onOverlayMouseLeave, overlay, title]);
  if (!(title !== null && title !== void 0 ? title : overlay)) {
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, children);
  }
  return /*#__PURE__*/_react.default.createElement(Popper, (0, _extends2.default)({
    ref: ref,
    prefixCls: cssScope(),
    popupPlacement: placement,
    action: trigger,
    popup,
    children: mergedChildren,
    // arrow
    showArrow,
    arrowHasStroke,
    // popupMotion
    popupMotion: 'tooltip-zoom',
    popupStyle: overlayStyle,
    popupClassName: overlayClassName,
    mouseLeaveDelay,
    defaultMainAxisOffsetWithoutArrow: 4
  }, restProps, {
    __screenshotAlwaysOpenEnabled: true
  }));
});
if (process.env.NODE_ENV !== "production") {
  Tooltip.displayName = 'Tooltip';
}
