// @universe-design/react/src/shared/components/list/index.stories.tsx
/* eslint-disable @typescript-eslint/no-magic-numbers */
/* eslint-disable react/react-in-jsx-scope */
/** @jsxImportSource @emotion/react */
import React from 'react';
import { Global } from '@emotion/react';
import styled from '@emotion/styled';
import AddOutlined from '@universe-design/icons-react/AddOutlined';
import CloseSmallOutlined from '@universe-design/icons-react/CloseSmallOutlined';
import MemberOutlined from '@universe-design/icons-react/MemberOutlined';
import RightBoldOutlined from '@universe-design/icons-react/RightBoldOutlined';
import List from '@universe-design/rc-virtual-list';
import Trigger from '@universe-design/rc-trigger';
import type { CheckboxProps } from '../../..';
import { Tag, Avatar, Loading, Button, Checkbox, Input, Space } from '@universe-design/react';
import type { ListItemProps } from '.';
import { ListItem, ListDivider, ListCaption, ListEmpty, ListItemLike, ListItemTitle, ListItemDesc } from '.';
import type { ExpandButtonProps } from '../expand-button';
import { ExpandButton } from '../expand-button';
import { ListBox } from './List';

export default {
  title: 'Shared Components/List',
};

const item = {
  title: 'title title title title title title title title',
  desc: 'desc desc desc desc desc desc desc desc',
  text: 'long long long long long long long long',
  src: 'https://s3-alpha-sig.figma.com/img/d945/a06b/efdeb2637342eb3d818bee30323ed17b?Expires=1638144000&Signature=UiwUWrY1xmX4FGFRMzBe3Q9MXmVLiMTxx2sn3qqSAv7wNnffYEmOl5ydvUUYW533yX72KDET3MMyV1VHjfy9JcBA9-TFGM1OFXSfpf8mQD7h0KNy9jzOKJOaTLHzbN6Lvn8OxgKePueNyWyWOTv7EbrAV6qVWJlm2I8kH~p1z9I5t5AV~ang7NS7m9xE7C4kf2zPUGjPIQ6LK0p1Eqk5E5dq1CTUxenGspGs1HhO0VfgaJvkncEAyPbysWcoksh-jfQ6PxJQOOlPpmrSxcewgEvtXQ9244xFIp1c7-OrHCnL3xgIbHx7lcykhM-CcM~IjEpnrIF-w828k4tXpzaJMQ__&Key-Pair-Id=APKAINTVSUGEWH5XD5UA',
};

const ListItemExpandIcon = styled(RightBoldOutlined)({
  fontSize: 12,
  marginLeft: 4,
  color: 'var(--icon-n2)',
});

const ListItems = (props: ListItemProps) => (
  <div
    css={{
      display: 'flex',
      flexDirection: 'column',
      width: 240,
      background: 'var(--bg-float)',
      padding: '2px 0',
      border: '1px solid var(--line-border-component)',
      borderRadius: 8,
      boxShadow: 'var(--shadow-s4-down)',
    }}
  >
    <ListCaption>Normal</ListCaption>
    <ListItem {...props} />
    <ListDivider />
    <ListItem {...props} />
    <ListItem {...props} selected />
    <ListDivider />
    <ListCaption>Disabled</ListCaption>
    <ListItem {...props} disabled />
    <ListItem {...props} disabled selected />
    <ListItem {...props} disabled>
      <ListItemTitle before={<Checkbox style={{ marginRight: 4 }} />}>normal-checkbox</ListItemTitle>
    </ListItem>
    <ListItem {...props} disabled>
      <ListItemTitle before={<Checkbox style={{ marginRight: 4 }} checked />}>checked checkbox</ListItemTitle>
    </ListItem>
  </div>
);

const EmbeddedList = () => (
  <List<number>
    data={Array<number>(10)
      .fill(0)
      .map((_, i) => i)}
    itemKey={o => o}
    height={32 * 7.5}
  >
    {o => {
      switch (item) {
        default:
          return (
            <ListItem>
              <ListItemTitle>{`选项 ${o}`}</ListItemTitle>
            </ListItem>
          );
      }
    }}
  </List>
);

const TreeBefore = ({
  depth,
  isExpanded,
  isLeaf = false,
}: {
  depth: number;
  isLeaf?: boolean;
} & Pick<ExpandButtonProps, 'isExpanded'>) => (
  <>
    {Array(depth)
      .fill(0)
      .map((_, i) => (
        <span key={i} css={{ width: 16 }} />
      ))}
    <ExpandButton
      css={{
        margin: '3px 3px 3px 0',
      }}
      iconType="ExpandDownFilled"
      isExpanded={isExpanded}
      asPlaceholder={isLeaf}
    />
  </>
);

export const All = () => (
  <>
    <Global
      styles={{
        body: {
          background: '#E5E5E5',
        },
      }}
    />
    <div
      css={{
        display: 'grid',
        gap: 16,
        gridTemplateColumns: 'repeat(7, 240px)',
        gridTemplateRows: 'repeat(6, 600px)',
      }}
    >
      {[true, false].map(textOverflow => (
        <>
          <div>
            <ListItems>
              <ListItemTitle textOverflow={textOverflow}>{item.title}</ListItemTitle>
            </ListItems>
          </div>
          <div>
            <ListItems showCheckmark={false}>
              <ListItemTitle textOverflow={textOverflow}>{item.title}</ListItemTitle>
            </ListItems>
          </div>
          <div>
            <ListItems>
              <ListItemTitle
                textOverflow={textOverflow}
                before={
                  <MemberOutlined
                    css={{
                      margin: '0 4px 0 0',
                    }}
                  />
                }
              >
                {item.title}
              </ListItemTitle>
            </ListItems>
          </div>
          <div>
            <ListItems>
              <ListItemTitle
                textOverflow={textOverflow}
                before={
                  <div
                    css={{
                      margin: '-1px 4px -1px 0',
                    }}
                  >
                    <Tag color="blue">标签</Tag>
                  </div>
                }
              >
                {item.title}
              </ListItemTitle>
            </ListItems>
          </div>
          <div>
            <ListItems>
              <ListItemTitle
                textOverflow={textOverflow}
                before={
                  <div
                    css={{
                      display: 'flex',
                      margin: '-1px 4px -1px 0',
                    }}
                  >
                    <Avatar src={item.src} size="small" color="blue">
                      头像
                    </Avatar>
                  </div>
                }
              >
                {item.title}
              </ListItemTitle>
            </ListItems>
          </div>
          <div>
            <ListItems>
              <ListItemTitle textOverflow={textOverflow}>{item.title}</ListItemTitle>
              <ListItemDesc>{item.desc}</ListItemDesc>
            </ListItems>
          </div>
          <div>
            <ListItems
              before={
                <Avatar
                  css={{
                    margin: '0 4px 0 0',
                    flex: 'none',
                    alignSelf: 'flex-start',
                  }}
                  src={item.src}
                  color="blue"
                >
                  头像
                </Avatar>
              }
            >
              <ListItemTitle textOverflow={textOverflow}>{item.title}</ListItemTitle>
              <ListItemDesc>{item.desc}</ListItemDesc>
            </ListItems>
          </div>
        </>
      ))}
      <div>
        <p>最简单的 select list</p>
        <ListBox>
          <EmbeddedList />
        </ListBox>
      </div>
      <div>
        <p>加载</p>
        <ListBox>
          <ListItemLike
            css={{
              justifyContent: 'center',
              alignItems: 'center',
            }}
          >
            <Loading size="small" />
            <div
              css={{
                height: 22,
              }}
            />
          </ListItemLike>
        </ListBox>
      </div>
      <div>
        <p>没有搜索结果</p>
        <ListBox>
          <ListEmpty>没有找到相关结果</ListEmpty>
        </ListBox>
      </div>
      <div>
        <p>初始无选项</p>
        <ListBox>
          <ListEmpty>暂无可选选项</ListEmpty>
        </ListBox>
      </div>
      <div>
        <p>弱网加载失败</p>
        <ListBox>
          <ListEmpty>
            选项加载失败
            <div
              css={{
                marginLeft: 12,
              }}
            >
              <div
                css={{
                  margin: '-2px -4px',
                }}
              >
                <Button type="text" color="primary">
                  重试
                </Button>
              </div>
            </div>
          </ListEmpty>
        </ListBox>
      </div>
      <div>
        <p>扩展选项 - 1</p>
        <ListBox>
          <ListItemLike>
            <div
              css={{
                margin: '-2px -4px',
              }}
            >
              <Button type="text" icon={<AddOutlined />} color="primary">
                新建「选项」
              </Button>
            </div>
          </ListItemLike>
          <EmbeddedList />
        </ListBox>
      </div>
      <div>
        <p>扩展选项 - 2</p>
        <ListBox>
          <EmbeddedList />
          <ListDivider fluid marginBlock="end" />
          <ListItemLike>
            <div
              css={{
                margin: '-2px -4px',
              }}
            >
              <Button type="text" icon={<AddOutlined />} color="primary">
                新建「选项」
              </Button>
            </div>
          </ListItemLike>
        </ListBox>
      </div>
      <div>
        <p>列表全选</p>
        <ListBox>
          <EmbeddedList />
          <ListDivider fluid marginBlock="end" />
          <ListItemLike>
            <Checkbox>全选（12）</Checkbox>
          </ListItemLike>
        </ListBox>
      </div>
      <div>
        <p>下拉列表包含搜索功能 - 1</p>
        <ListBox>
          <Input.Search
            css={{
              borderColor: 'transparent !important',
              borderRadius: 0,
            }}
            allowClear
            placeholder="请输入文字"
          />
          <ListDivider fluid marginBlock="start" />
          <EmbeddedList />
        </ListBox>
      </div>
      <div>
        <p>下拉列表包含搜索功能 - 2</p>
        <ListBox>
          <Input.Search
            css={{
              borderColor: 'transparent !important',
              borderRadius: 0,
            }}
            allowClear
            placeholder="请输入文字"
            defaultValue="选"
          />
          <ListDivider fluid marginBlock="start" />
          <ListEmpty>没有找到相关结果</ListEmpty>
        </ListBox>
      </div>
      <div>
        <p>下拉菜单</p>
        <ListBox
          style={{
            width: 148,
          }}
        >
          <ListItem>
            <ListItemTitle>操作</ListItemTitle>
          </ListItem>
          <ListDivider fluid />
          <ListItem selected>
            <ListItemTitle>操作</ListItemTitle>
          </ListItem>
          <ListItem selected>
            <ListItemTitle>操作</ListItemTitle>
            <ListItemDesc>辅助信息</ListItemDesc>
          </ListItem>
          <ListItem selected disabled>
            <ListItemTitle>操作</ListItemTitle>
            <ListItemDesc>辅助信息</ListItemDesc>
          </ListItem>
          <ListDivider fluid />
          <ListCaption>标题</ListCaption>
          <Trigger
            popupVisible
            popup={
              <ListBox
                style={{
                  width: 148,
                }}
              >
                <ListItem active selected>
                  <ListItemTitle>操作</ListItemTitle>
                </ListItem>
                <ListItem>
                  <ListItemTitle>操作</ListItemTitle>
                </ListItem>
                <ListItem>
                  <ListItemTitle>操作</ListItemTitle>
                </ListItem>
              </ListBox>
            }
            popupAlign={{
              points: ['tl', 'tr'],
              offset: [3, -4],
            }}
          >
            <ListItem after={<ListItemExpandIcon />} active>
              <ListItemTitle>操作</ListItemTitle>
            </ListItem>
          </Trigger>
          <ListItem after={<ListItemExpandIcon />}>
            <ListItemTitle>操作</ListItemTitle>
            <ListItemDesc>辅助信息</ListItemDesc>
          </ListItem>
          <ListItem after={<ListItemExpandIcon />} disabled>
            <ListItemTitle>操作</ListItemTitle>
          </ListItem>
        </ListBox>
      </div>
      <div />
      {[true, false].map(isSelect => (
        <div key={String(isSelect)}>
          <p>树选择</p>
          <ListBox
            css={{
              '& .ud__checkbox__wrapper': {
                marginRight: 8,
              },
            }}
          >
            <ListItem>
              <ListItemTitle
                before={
                  <>
                    <TreeBefore depth={0} isExpanded={false} />
                    {!isSelect && <Checkbox indeterminate />}
                  </>
                }
              >
                一级 1
              </ListItemTitle>
            </ListItem>
            <ListItem>
              <ListItemTitle
                before={
                  <>
                    <TreeBefore depth={0} isExpanded={true} />
                    {!isSelect && <Checkbox indeterminate />}
                  </>
                }
              >
                一级 2
              </ListItemTitle>
            </ListItem>
            <ListItem>
              <ListItemTitle
                before={
                  <>
                    <TreeBefore depth={1} isExpanded={true} />
                    {!isSelect && <Checkbox indeterminate />}
                  </>
                }
              >
                二级 2-1
              </ListItemTitle>
            </ListItem>
            <ListItem>
              <ListItemTitle
                before={
                  <>
                    <TreeBefore depth={2} isExpanded={false} />
                    {!isSelect && <Checkbox />}
                  </>
                }
              >
                三级 2-1-1
              </ListItemTitle>
            </ListItem>
            <ListItem selected={isSelect}>
              <ListItemTitle
                before={
                  <>
                    <TreeBefore depth={2} isExpanded={false} isLeaf />
                    {!isSelect && <Checkbox checked />}
                  </>
                }
              >
                三级 2-1-2
              </ListItemTitle>
            </ListItem>
            <ListItem selected={isSelect}>
              <ListItemTitle
                before={
                  <>
                    <TreeBefore depth={2} isExpanded={false} isLeaf />
                    {!isSelect && <Checkbox checked />}
                  </>
                }
              >
                三级 2-1-3
              </ListItemTitle>
            </ListItem>
            <ListItem>
              <ListItemTitle
                before={
                  <>
                    <TreeBefore depth={1} isExpanded={false} />
                    {!isSelect && <Checkbox />}
                  </>
                }
              >
                二级 2-2
              </ListItemTitle>
            </ListItem>
          </ListBox>
        </div>
      ))}
    </div>
  </>
);

export const Disabled = () => {
  const CheckboxBefore = (checkboxProps: CheckboxProps) => (
    <div style={{ marginRight: 4 }}>
      <Checkbox {...checkboxProps} />
    </div>
  );

  return (
    <Space>
      {([true, false] as boolean[]).map(isDisabled => (
        <ListBox style={{ width: 300 }} key={String(isDisabled)}>
          <ListItem disabled={isDisabled} before={<CheckboxBefore />}>
            <ListItemTitle>normal checkbox</ListItemTitle>
          </ListItem>
          <ListItem disabled={isDisabled} before={<CheckboxBefore disabled />}>
            <ListItemTitle>disabled checkbox</ListItemTitle>
          </ListItem>
          <ListItem disabled={isDisabled} before={<CheckboxBefore checked />}>
            <ListItemTitle before={<CheckboxBefore checked />}>checked-checkbox</ListItemTitle>
          </ListItem>
          <ListItem before={<CheckboxBefore checked disabled />} disabled={isDisabled}>
            <ListItemTitle>disabled-checked-checkbox</ListItemTitle>
          </ListItem>
          <ListItem disabled={isDisabled}>
            <ListItemTitle>
              primary-text-button
              <Button type="text" color="primary">
                Text Button
              </Button>
            </ListItemTitle>
          </ListItem>
          <ListItem disabled={isDisabled}>
            <ListItemTitle>
              text-button{' '}
              <Button type="text" color="primary" disabled>
                {' '}
                disabled Text Button
              </Button>
            </ListItemTitle>
          </ListItem>
          <ListItem disabled={isDisabled}>
            <ListItemTitle
              before={
                <div
                  css={{
                    display: 'flex',
                    margin: '-1px 4px -1px 0',
                  }}
                >
                  <Avatar
                    src="https://s3-alpha-sig.figma.com/img/d945/a06b/efdeb2637342eb3d818bee30323ed17b?Expires=1638144000&Signature=UiwUWrY1xmX4FGFRMzBe3Q9MXmVLiMTxx2sn3qqSAv7wNnffYEmOl5ydvUUYW533yX72KDET3MMyV1VHjfy9JcBA9-TFGM1OFXSfpf8mQD7h0KNy9jzOKJOaTLHzbN6Lvn8OxgKePueNyWyWOTv7EbrAV6qVWJlm2I8kH~p1z9I5t5AV~ang7NS7m9xE7C4kf2zPUGjPIQ6LK0p1Eqk5E5dq1CTUxenGspGs1HhO0VfgaJvkncEAyPbysWcoksh-jfQ6PxJQOOlPpmrSxcewgEvtXQ9244xFIp1c7-OrHCnL3xgIbHx7lcykhM-CcM~IjEpnrIF-w828k4tXpzaJMQ__&Key-Pair-Id=APKAINTVSUGEWH5XD5UA"
                    size="small"
                    color="blue"
                  >
                    头像
                  </Avatar>
                </div>
              }
            >
              Option with Avatar
            </ListItemTitle>
            <ListItemDesc>item description</ListItemDesc>
          </ListItem>
          <ListItem disabled={isDisabled}>
            <ListItemTitle>
              Option with Tag
              <Tag size="small" color="blue">
                标签
              </Tag>
            </ListItemTitle>
            <ListItemDesc>item description</ListItemDesc>
          </ListItem>
          <ListItem disabled={isDisabled} after={<Button type="icon" size="small" icon={<CloseSmallOutlined />} />}>
            <ListItemTitle>Option with close icon button</ListItemTitle>
            <ListItemDesc>item description</ListItemDesc>
          </ListItem>
          <ListItem
            disabled={isDisabled}
            after={<Button type="icon" size="small" icon={<CloseSmallOutlined />} disabled />}
          >
            <ListItemTitle>Option with disabled close icon button</ListItemTitle>
            <ListItemDesc>item description</ListItemDesc>
          </ListItem>
          <ListItem disabled={isDisabled} selected>
            <ListItemTitle>selected option</ListItemTitle>
            <ListItemDesc>item description</ListItemDesc>
          </ListItem>
        </ListBox>
      ))}
    </Space>
  );
};
