// @universe-design/react/src/shared/components/list/ListItem.tsx
import ListCheckBoldOutlined from '@universe-design/icons-react/ListCheckBoldOutlined';
import clsx from 'clsx';
import React from 'react';
import stylex from '@universe-design/stylex';
import { cssTypographyBody2, cssFlexBase, cssTypographyHeadline, cssTypographyBody0 } from '../../../stylex';
import type { StylexProps } from '../../../util';
import { isNonNullable } from '../../../util';
import { useCssScope } from '../../../config-provider/hooks';
import { TextOverflow } from '../text-overflow';

import './index.less';

const cssListItemLike = stylex.compose(cssFlexBase, cssTypographyBody0);

export type ListItemLikeProps = React.ComponentPropsWithoutRef<'div'>;

export const ListItemLike = React.forwardRef<HTMLDivElement, ListItemLikeProps>(({ className, ...restProps }, ref) => {
  const cssScope = useCssScope('list__itemLike');

  return <div ref={ref} className={clsx(cssScope(), stylex(cssListItemLike), className)} {...restProps} />;
});

export interface ListItemProps extends React.ComponentPropsWithoutRef<'li'>, StylexProps {
  disabled?: boolean;
  selected?: boolean;
  active?: boolean;
  before?: React.ReactNode;
  after?: React.ReactNode;
  showCheckmark?: boolean;
}

export const ListItem = React.forwardRef<HTMLLIElement, ListItemProps>(
  (
    {
      className,
      disabled = false,
      selected = false,
      active = false,
      before,
      after,
      children,
      showCheckmark = true,
      sx,
      ...restProps
    },
    ref,
  ) => {
    const cssScope = useCssScope('list__item');
    const itemLikeCssScope = useCssScope('list__itemLike');

    return (
      <li
        ref={ref}
        className={clsx(
          className,
          cssScope(),
          itemLikeCssScope(),
          stylex(cssListItemLike, ...(sx ?? [])),
          disabled && cssScope('-disabled'),
          selected && cssScope('-selected'),
          active && cssScope('-active'),
        )}
        {...restProps}
      >
        {isNonNullable(before) && <div className={cssScope('__before')}>{before}</div>}
        {isNonNullable(children) && <div className={cssScope('__content')}>{children}</div>}
        {isNonNullable(after) ? (
          <div className={cssScope('__after')}>{after}</div>
        ) : showCheckmark ? (
          <ListCheckBoldOutlined className={clsx(cssScope('__checkmark'))} />
        ) : null}
      </li>
    );
  },
);

export interface ListItemTitleProps extends React.ComponentPropsWithoutRef<'div'> {
  before?: React.ReactNode;
  textOverflow?: boolean;
}

export const ListItemTitle = React.forwardRef<HTMLDivElement, ListItemTitleProps>(
  ({ className, children, before, textOverflow = true, ...restProps }, ref) => {
    const cssScope = useCssScope('list__item__title');

    return (
      <div ref={ref} className={clsx(cssScope(), className)} {...restProps}>
        {isNonNullable(before) && <div className={cssScope('__before')}>{before}</div>}
        {isNonNullable(children) &&
          (textOverflow ? (
            <TextOverflow as="div" lineClamp={1}>
              {children}
            </TextOverflow>
          ) : (
            children
          ))}
      </div>
    );
  },
);

export type ListItemDescProps = React.ComponentPropsWithoutRef<'div'>;

export const ListItemDesc = React.forwardRef<HTMLDivElement, ListItemDescProps>(({ className, ...restProps }, ref) => {
  const cssScope = useCssScope('list__item__description');

  return <div ref={ref} className={clsx(cssScope(), stylex(cssTypographyBody2), className)} {...restProps} />;
});

export interface ListDividerProps extends React.ComponentPropsWithoutRef<'hr'> {
  /**
   * 是否占满宽度
   * @defaultValue false
   */
  fluid?: boolean;
  /**
   * 上下 margin 的模式
   * @defaultValue `both`
   */
  marginBlock?: 'start' | 'end' | 'both';
}

export const ListDivider = React.forwardRef<HTMLHRElement, ListDividerProps>(
  ({ className, fluid = false, marginBlock = 'both', ...restProps }, ref) => {
    const cssScope = useCssScope('list__divider');

    return (
      <hr
        ref={ref}
        className={clsx(cssScope(), className, fluid && cssScope('--fluid'), cssScope(`--margin-${marginBlock}`))}
        {...restProps}
      />
    );
  },
);

export type ListCaptionProps = React.ComponentPropsWithoutRef<'div'> & StylexProps;

export const ListCaption = React.forwardRef<HTMLDivElement, ListCaptionProps>(
  ({ className, sx, ...restProps }, ref) => {
    const cssScope = useCssScope('list__caption');
    const itemLikeCssScope = useCssScope('list__itemLike');

    return (
      <div
        ref={ref}
        className={clsx(
          cssScope(),
          itemLikeCssScope(),
          stylex(cssListItemLike, cssTypographyHeadline, ...(sx || [])),
          className,
        )}
        {...restProps}
      />
    );
  },
);

export type ListEmptyProps = React.ComponentPropsWithoutRef<'div'>;

export const ListEmpty = React.forwardRef<HTMLDivElement, ListEmptyProps>(({ className, ...restProps }, ref) => {
  const cssScope = useCssScope('list__empty');
  const itemLikeCssScope = useCssScope('list__itemLike');

  return (
    <div
      ref={ref}
      className={clsx(cssScope(), itemLikeCssScope(), stylex(cssListItemLike), className)}
      {...restProps}
    />
  );
});
