// @universe-design/react/src/popover/stories/homeless.stories.tsx
/** @jsxImportSource @emotion/react */
import * as React from 'react';
import DeleteTrashOutlined from '@universe-design/icons-react/DeleteTrashOutlined';
import type { DefaultSizeType, TooltipPlacement } from '../..';
import { Slider, Space, Button, Radio, Tooltip, Popover, Select } from '../..';
import './index.less';
import { default as placementDemo } from '../demo/placement';
import { PerformanceContent } from '../../_stories/common/PerformanceContent';
import { LogsContainer } from '../../_stories/common/LogsContainer';

export default {
  title: 'Components/Popover/临时示例',
  component: Popover,
};

export const BasicUsage = Object.assign((props: React.ComponentProps<typeof Popover>) => <Popover {...props} />);

export const placements = () => <div style={{ marginTop: 200, marginLeft: 200 }}>{placementDemo()}</div>;

export const Icons = () => (
  <Space size="large" wrap>
    <Popover title="Are you sure？" icon="success">
      <Button type="text">Success</Button>
    </Popover>
    <Popover title="Are you sure？" icon="warning">
      <Button type="text">Warning</Button>
    </Popover>
    <Popover title="Are you sure？" icon="info">
      <Button type="text">Info</Button>
    </Popover>
    <Popover title="Are you sure？" icon="error">
      <Button type="text" color="danger">
        Error
      </Button>
    </Popover>
    <Popover title="Are you sure？" icon={<DeleteTrashOutlined style={{ color: 'red' }} />} trigger="click">
      <Button type="text" color="danger">
        Delete (custom icon)
      </Button>
    </Popover>
  </Space>
);

const des = 'popover content. popover content. popover content. ';

export const Mode = () => (
  <Space size="large" wrap>
    <Popover title="Are you sure？" content={des} mode="normal" trigger="click">
      <Button type="text" href="#">
        normal
      </Button>
    </Popover>
    <Popover title="Are you sure？" content={des} mode="loose" trigger="click">
      <Button type="text" href="#">
        loose
      </Button>
    </Popover>
    <Popover title="Are you sure？" content={des} mode="pure" trigger="click">
      <Button type="text" href="#">
        pure
      </Button>
    </Popover>
  </Space>
);

export const ContentScrolled = () => {
  const renderContent = () => (
    <div className="ud__popover-content-scroll">
      <div>
        <span>内容区域</span>
      </div>
    </div>
  );

  return (
    <Space size="large">
      <Popover mode="pure" content={renderContent} trigger="click">
        <Button type="primary">Open</Button>
      </Popover>
      <Popover content={renderContent} trigger="click" mode="pure">
        <Button type="primary">无 padding</Button>
      </Popover>
    </Space>
  );
};

export const Disabled = () => (
  <Popover title="Are you sure？" disabled>
    <Button type="primary">禁用</Button>
  </Popover>
);

export const MaxWidth = () => (
  <Popover
    title="This is a Title"
    content="Only used for calendar details cards, extreme text prompt components for pop-up details."
  >
    <Button type="primary">默认最大长度</Button>
  </Popover>
);

const largeDes =
  '超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案';
const smallDes = '一段文案，一段文案';

export const PlacementCustomer = () => {
  const [marginTop, setMarginTop] = React.useState(100);
  const [marginLeft, setMarginLeft] = React.useState(100);
  const [placement, setPlacement] = React.useState<TooltipPlacement>('top');
  const [strSize, setStrSize] = React.useState<DefaultSizeType>('small');

  return (
    <div style={{ height: '200vh' }}>
      <Space size="large" direction="vertical" style={{ width: '100%' }}>
        <Slider step={100} min={0} max={1000} value={marginTop} onChange={setMarginTop} />
        <Slider step={100} min={0} max={1000} value={marginLeft} onChange={setMarginLeft} />
        <Space align="center">
          <span>方向:</span>
          <Radio.Group
            value={placement}
            onChange={v => {
              setPlacement(v as TooltipPlacement);
              console.log('placementState change', v);
            }}
          >
            <Radio.Button value="top">上</Radio.Button>
            <Radio.Button value="bottom">下</Radio.Button>
            <Radio.Button value="left">左</Radio.Button>
            <Radio.Button value="right">右</Radio.Button>
            <Radio.Button value="topLeft">上左</Radio.Button>
            <Radio.Button value="topRight">上右</Radio.Button>
            <Radio.Button value="bottomLeft">下左</Radio.Button>
            <Radio.Button value="bottomRight">下右</Radio.Button>
            <Radio.Button value="leftTop">左上</Radio.Button>
            <Radio.Button value="leftBottom">左下</Radio.Button>
            <Radio.Button value="rightTop">右上</Radio.Button>
            <Radio.Button value="rightBottom">右下</Radio.Button>
          </Radio.Group>
        </Space>
        <Space align="center">
          <span>内容大小:</span>
          <Radio.Group value={strSize} onChange={v => setStrSize(v as DefaultSizeType)}>
            <Radio.Button value="small">短内容</Radio.Button>
            <Radio.Button value="large">长内容</Radio.Button>
          </Radio.Group>
        </Space>
      </Space>
      <Popover
        placement={placement}
        title={strSize === 'small' ? smallDes : largeDes}
        onOverlayMouseEnter={() => console.log('onOverlayMouseEnter')}
        onOverlayMouseLeave={() => console.log('onOverlayMouseLeave')}
      >
        <Button style={{ marginTop, marginLeft }}>超长文案超长文案超长文案</Button>
      </Popover>
    </div>
  );
};

export const PopoverTooltip = () => (
  <Tooltip overlay="tooltip overlay">
    <Popover content="popover content">
      <Tooltip overlay="tooltip2 overlay">
        <Button>button</Button>
      </Tooltip>
    </Popover>
  </Tooltip>
);

export const PopoverForceRender = () => (
  <>
    <Popover
      trigger="click"
      content={
        <PerformanceContent count={5 * 100} componentName="PopoverForceRender" itemRenderer={() => <Select />} />
      }
      destroyOnHide={false}
    >
      <Button>button</Button>
    </Popover>
    <LogsContainer />
  </>
);
