// @universe-design/react/src/popover/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C5PlacementFallback from '../demo/5.placement-fallback';
import CBasic from '../demo/basic';
import CControl from '../demo/control';
import CPlacement from '../demo/placement';
import CTriggerType from '../demo/triggerType';
if (!(C5PlacementFallback as Story).storyName) {
  (C5PlacementFallback as Story).storyName = '5. 上下左右位置回退';
  (C5PlacementFallback as Story).parameters = {
    docs: {
      description: {
        story:
          '默认只会做对位翻转，也就是 `top` 对应 `bottom`，`left` 对应 `right`，如果需要其他的翻转模式可以通过 `flip` 属性设置。\n\n',
      },
    },
  };
}
if (!(CBasic as Story).storyName) {
  (CBasic as Story).storyName = '0. 基础用法';
  (CBasic as Story).parameters = {
    docs: {
      description: {
        story: '简单的用法，浮层的大小由内容区域决定。\n\n',
      },
    },
  };
}
if (!(CControl as Story).storyName) {
  (CControl as Story).storyName = '3. 从浮层内关闭';
  (CControl as Story).parameters = {
    docs: {
      description: {
        story: '使用 `visible` 属性控制浮层显示。\n\n',
      },
    },
  };
}
if (!(CPlacement as Story).storyName) {
  (CPlacement as Story).storyName = '2. 位置';
  (CPlacement as Story).parameters = {
    docs: {
      description: {
        story: '支持 12 个不同的方位。\n\n',
      },
    },
  };
}
if (!(CTriggerType as Story).storyName) {
  (CTriggerType as Story).storyName = '1. 三种触发方式';
  (CTriggerType as Story).parameters = {
    docs: {
      description: {
        story: '鼠标移入、聚集、点击。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Popover/官网示例',
  tags: ['autodocs'],
};

export { C5PlacementFallback, CBasic, CControl, CPlacement, CTriggerType };
