// @byted-service/logger/lib/streams/console.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.createConsole = void 0;
const tslib_1 = require("tslib");
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const colorizer = {
    fatal: chalk_1.default.magenta,
    error: chalk_1.default.red,
    warn: chalk_1.default.yellow,
    info: chalk_1.default.green,
    debug: chalk_1.default.blue,
    trace: chalk_1.default.cyan,
};
let stdout;
/* istanbul ignore next */
if (typeof ((_a = process.stdout) === null || _a === void 0 ? void 0 : _a.write) === 'function') {
    stdout = {
        write: process.stdout.write.bind(process.stdout),
    };
}
else {
    console.warn('Unsupport operating environment, process.stdout must be a tty or stream. \n' +
        "Now @byted-service/logger's console output just use console.log().");
    stdout = {
        write: console.log.bind(console),
    };
}
const createConsole = () => {
    return {
        stdout,
        write(msg, inputData) {
            const levelLen = inputData.level.length;
            const color = colorizer[inputData.level] || colorizer.info;
            stdout.write(`${color(msg.slice(0, levelLen))}${msg.slice(levelLen)}`);
        },
    };
};
exports.createConsole = createConsole;
