// @byted-service/logger/lib/streams/logagent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createLogAgent = void 0;
const env_1 = require("@byted-service/env");
const logagent_1 = require("@byted-service/logagent");
const logger_1 = require("../logger");
let defaultLogger = null;
const getLogger = () => {
    return (defaultLogger =
        defaultLogger ||
            new logger_1.Logger({
                enableAgentLog: false,
                enableMetrics: false,
            }));
};
const tagNames = [
    'called',
    'rip',
    'from_method',
    'method',
    'cost',
    'status',
    'stress_tag',
    'node_type',
    'to_cluster',
    'full_path',
    'from_cluster',
    '_psm',
    '_ip',
    '_cluster',
    '_podName',
    '_host',
    '_version',
    '_deployStage',
];
const parseMsg = (message, joiner = ' ') => {
    const tags = {};
    const tagKeys = [];
    const pureMsgArr = [];
    message.split(joiner).forEach((tag) => {
        const [k, v] = tag.split(/=(.*)/);
        if (v && tagNames.includes(k)) {
            tags[k] = v;
            tagKeys.push(k);
        }
        else {
            pureMsgArr.push(tag);
        }
    });
    return {
        tags,
        msg: pureMsgArr.join(joiner),
        tagKeys,
    };
};
const createLogAgent = (options = {}) => {
    const logAgent = new logagent_1.LogagentV2(Object.assign(Object.assign({ logger: getLogger() }, options.agent), { psm: options.taskName || options.psm || (0, env_1.getPSM)() }));
    return {
        logAgent,
        write(_, inputData) {
            let msg;
            let tags;
            let tagKeys;
            if (inputData.tags) {
                msg = inputData.message;
                tags = inputData.tags;
                tagKeys = inputData.tagKeys;
            }
            else {
                const result = parseMsg(inputData.message);
                msg = result.msg;
                tags = result.tags;
                tagKeys = result.tagKeys;
            }
            tags._level = inputData.level.toUpperCase();
            tags._ts = inputData.ts;
            tags._logid = inputData.logId;
            tags._location = inputData.location;
            tags._version = inputData.version;
            tags._psm = inputData.psm;
            tags._cluster = inputData.cluster;
            tags._idc = inputData.idc;
            tags._deployStage = inputData.deployStage;
            logAgent.send(msg, tags, tagKeys);
        },
        flush() {
            logAgent.flush();
        },
    };
};
exports.createLogAgent = createLogAgent;
