// @byted-service/logger/lib/streams/metrics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createMetrics = void 0;
const tslib_1 = require("tslib");
const metrics_1 = tslib_1.__importDefault(require("@byted-service/metrics"));
const logger_1 = require("../logger");
const consts_1 = require("../consts");
let defaultLogger = null;
const getLogger = () => {
    return (defaultLogger =
        defaultLogger ||
            new logger_1.Logger({
                enableAgentLog: false,
                enableMetrics: false,
            }));
};
const createMetrics = (options = {}) => {
    const metrics = options.metrics instanceof metrics_1.default
        ? options.metrics
        : new metrics_1.default(Object.assign({ defaultPrefix: 'toutiao.service.log', logger: getLogger() }, options.metrics));
    return {
        metrics,
        write(_, inputData) {
            const { level, psm, cluster, env } = inputData;
            metrics.emitCounter(`${psm}.throughput`, 1, '', {
                level: level.toUpperCase(),
                cluster,
                env,
                version: consts_1.version,
            });
        },
        flush() {
            metrics.flush().catch(consts_1.noop);
        },
    };
};
exports.createMetrics = createMetrics;
