// @universe-design/react/lib/radio/group.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _reactUtilities = require("@universe-design/react-utilities");
var _radio = _interopRequireDefault(require("./radio"));
var _hooks = require("../config-provider/hooks");
var _context = require("./context");
var _util = require("../util");
require("./style");
/**
 * @public
 */

/**
 * @public
 */

/**
 * 单选框组合，用于包裹一组单选器
 */

/**
 * 如果有 radio button child，父级 optionType 就要是 button
 * https://stackoverflow.com/questions/21252551/apply-style-to-parent-if-it-has-child-with-css
 */
function useRadioGroupButtonOptionType(_ref) {
  let {
    optionType: optionTypeProp = 'default'
  } = _ref;
  const [optionTypeRegistry, setButtonRegistry] = _react.default.useState(new Map());
  const registerBOT = _react.default.useCallback(value => {
    setButtonRegistry(s => new Map([...s.entries(), [value, 'button']]));
  }, []);
  const unregisterBOT = _react.default.useCallback(value => {
    setButtonRegistry(s => {
      s.delete(value);
      return new Map([...s.entries()]);
    });
  }, []);
  const optionType = optionTypeRegistry.size ? 'button' : optionTypeProp;
  return {
    registerBOT,
    unregisterBOT,
    optionType
  };
}
const RadioGroup = /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
  const {
    className,
    options,
    direction = 'horizontal',
    // buttonStyle = 'outline',
    disabled,
    children,
    size,
    style,
    id,
    onMouseEnter,
    onMouseLeave,
    prefixCls
  } = props;
  const ctxComponentSize = (0, _hooks.useComponentSize)();
  const mergedSize = size !== null && size !== void 0 ? size : ctxComponentSize;
  const groupCssScope = (0, _hooks.useCssScope)('radio-group', prefixCls);
  const radioButtonScope = (0, _hooks.useCssScope)('radio-button', prefixCls)();
  const [value, setValue] = (0, _reactUtilities.useControlledState)(props.defaultValue, props.value, {
    isControlledOverride: 'value' in props
  });
  const {
    registerBOT,
    unregisterBOT,
    optionType
  } = useRadioGroupButtonOptionType(props);
  return /*#__PURE__*/_react.default.createElement(_context.RadioGroupContextProvider, {
    value: {
      onChange: ev => {
        const val = ev.target.value;
        // 在 group 模式下，如果没有提供 val，不会触发 onChange。
        if (val === undefined) {
          return;
        }
        setValue(val);
        const {
          onChange
        } = props;
        (0, _util.isFunction)(onChange) && val !== value && onChange(val, ev);
      },
      value,
      disabled: props.disabled,
      name: props.name,
      error: props.error,
      registerBOT,
      unregisterBOT,
      size: mergedSize
    }
  }, (() => {
    let renderRadio = children;
    if (options && options.length > 0) {
      const radioPrefixCls = optionType === 'button' ? radioButtonScope : undefined;
      renderRadio = options.map(option => {
        if (typeof option === 'string') {
          return /*#__PURE__*/_react.default.createElement(_radio.default, {
            key: option,
            disabled: disabled,
            value: option,
            checked: value === option,
            prefixCls: radioPrefixCls,
            optionType: optionType
          }, option);
        }
        return /*#__PURE__*/_react.default.createElement(_radio.default, {
          key: `radio-option-${option.value}`,
          disabled: option.disabled || disabled,
          value: option.value,
          checked: value === option.value,
          style: option.style,
          className: option.className,
          prefixCls: radioPrefixCls,
          optionType: optionType
        }, option.label);
      });
    }
    const componentSize = (0, _util.getDefaultSizeAbbr)(mergedSize);
    const classNames = (0, _clsx.default)(groupCssScope(), optionType === 'button' && groupCssScope('-button'), groupCssScope('--mode-solid'),
    // 目前我们的设计只有 solid mode 一种样式
    groupCssScope(`--direction-${direction}`), {
      [groupCssScope(`--size-${componentSize}`)]: componentSize
    }, className);
    return /*#__PURE__*/_react.default.createElement("div", {
      className: classNames,
      style: style,
      onMouseEnter: onMouseEnter,
      onMouseLeave: onMouseLeave,
      id: id,
      ref: ref
    }, renderRadio);
  })());
});
var _default = exports.default = /*#__PURE__*/_react.default.memo(RadioGroup);
