// @universe-design/react/lib/radio/radio.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Radio = void 0;
var _u5msrnt = _interopRequireDefault(require("./../.stylex/u5msrnt"));
var _u1pdmi = _interopRequireDefault(require("./../.stylex/u1pdmi33"));
var _u1tqlmkb = _interopRequireDefault(require("./../.stylex/u1tqlmkb"));
var _u15pdtm = _interopRequireDefault(require("./../.stylex/u15pdtm"));
var _u143d86t = _interopRequireDefault(require("./../.stylex/u143d86t"));
var _u1y2b6fc = _interopRequireDefault(require("./../.stylex/u1y2b6fc"));
var _urjm3ap = _interopRequireDefault(require("./../.stylex/urjm3ap"));
var _u1rs2opo = _interopRequireDefault(require("./../.stylex/u1rs2opo"));
var _uh8azbh = _interopRequireDefault(require("./../.stylex/uh8azbh"));
var _upgelb = _interopRequireDefault(require("./../.stylex/upgelb9"));
var _u441rjq = _interopRequireDefault(require("./../.stylex/u441rjq"));
var _u1vsjzop = _interopRequireDefault(require("./../.stylex/u1vsjzop"));
var _um3isuf = _interopRequireDefault(require("./../.stylex/um3isuf"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _reactUtilities = require("@universe-design/react-utilities");
var _stylex2 = require("../stylex");
var _useRadioGroupContext = _interopRequireDefault(require("./useRadioGroupContext"));
var _hooks = require("../config-provider/hooks");
var _util = require("../util");
var _text = require("../text");
require("./style");
//.uf9s3yw{margin-left:8px;}
//.uokg9q4{overflow-wrap:break-word;}
//.u132xexn{max-width:240px;}
//.u1sbtcvk{padding-top:5px;}
//.u11qrl6u{padding-right:12px;}
//.udghr9{padding-bottom:5px;}
//.ujlbh76{padding-left:12px;}
//.u1sl3k7w{padding-top:1px;}
//.u1brlhvm{padding-bottom:1px;}
//.uq1loh5{padding-top:2px;}
//.u19lj068{padding-right:8px;}
//.uutqtb8{padding-bottom:2px;}
//.u177v4lu{padding-left:8px;}
const RADIO_LABEL_MARGIN = 8;
const RADIO_BUTTON_MAX_WIDTH = 240;
const Radio = exports.Radio = /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
  const {
    prefixCls,
    className,
    role = 'radio',
    checked,
    style,
    onChange,
    onMouseEnter,
    onMouseLeave,
    children,
    value,
    disabled,
    name,
    defaultChecked,
    error,
    optionType = 'default',
    ...restProps
  } = props;
  const context = (0, _useRadioGroupContext.default)();
  const inputRef = (0, _react.useRef)(null);
  const cssScope = (0, _hooks.useCssScope)('radio', prefixCls);
  const rootPrefixCls = cssScope();
  const isButton = optionType === 'button';
  const [checkedState, setCheckedState] = (0, _reactUtilities.useControlledState)(Boolean(defaultChecked), context ? context.value === value : Boolean(checked), {
    isControlledOverride: Boolean(context) || 'checked' in props
  });
  (0, _react.useImperativeHandle)(ref, () => inputRef.current);
  let disableState = disabled;
  let nameState = name;
  let errorState = error;
  if (context) {
    disableState = disabled || context.disabled;
    nameState = context.name;
    errorState = error || context.error;
  }
  const wrapperCls = (() => {
    let cssWrapperSize;
    if ((context === null || context === void 0 ? void 0 : context.size) === 'small') {
      cssWrapperSize = {
        "props": {
          "z8tnut": _upgelb.default,
          "z189sj": _u441rjq.default,
          "Byoj8tv": _u1vsjzop.default,
          "uwmqm3": _um3isuf.default
        }
      };
    } else if ((context === null || context === void 0 ? void 0 : context.size) === 'medium') {
      cssWrapperSize = {
        "props": {
          "z8tnut": _u1rs2opo.default,
          "z189sj": _u441rjq.default,
          "Byoj8tv": _uh8azbh.default,
          "uwmqm3": _um3isuf.default
        }
      };
    } else if ((context === null || context === void 0 ? void 0 : context.size) === 'large') {
      cssWrapperSize = {
        "props": {
          "z8tnut": _u15pdtm.default,
          "z189sj": _u143d86t.default,
          "Byoj8tv": _u1y2b6fc.default,
          "uwmqm3": _urjm3ap.default
        }
      };
    }
    return (0, _clsx.default)(`${rootPrefixCls}__wrapper`, {
      [`${rootPrefixCls}__wrapper--checked`]: checkedState,
      [`${rootPrefixCls}__wrapper--disable`]: disableState,
      [`${rootPrefixCls}--error`]: errorState
    }, className, (0, _stylex.default)(_stylex2.cssTypographyBody0,
    // 使用 button style 的时候选中需要加粗
    isButton && checkedState && _stylex2.cssFontMedium, isButton && cssWrapperSize));
  })();
  const sxLabelContent = (() => {
    let cssContentSize;
    if ((context === null || context === void 0 ? void 0 : context.size) === 'small') {
      cssContentSize = _stylex2.cssTypographyBody2;
    } else if ((context === null || context === void 0 ? void 0 : context.size) === 'medium') {
      cssContentSize = _stylex2.cssTypographyBody0;
    } else if ((context === null || context === void 0 ? void 0 : context.size) === 'large') {
      cssContentSize = _stylex2.cssTypographyBody0;
    }
    const sxArray = [_stylex2.cssBlockBase];
    if (isButton) {
      sxArray.push(cssContentSize, {
        "props": {
          "B2u0y6b": _u1tqlmkb.default
        }
      });
    } else {
      sxArray.push({
        "props": {
          "Frg6f3": _u5msrnt.default,
          "Btd35i7": _u1pdmi.default
        }
      });
    }
    return sxArray;
  })();
  return /*#__PURE__*/_react.default.createElement("label", {
    className: wrapperCls,
    style: style,
    onMouseEnter: onMouseEnter,
    onMouseLeave: onMouseLeave
  }, /*#__PURE__*/_react.default.createElement("span", {
    className: rootPrefixCls
  }, /*#__PURE__*/_react.default.createElement("input", (0, _extends2.default)({
    ref: inputRef,
    className: (0, _clsx.default)(cssScope('__input')),
    type: "radio",
    role: role,
    checked: checkedState,
    onChange: function (e) {
      const eventParams = {
        target: {
          type: 'radio',
          checked: e.target.checked,
          value
        },
        stopPropagation() {
          e.stopPropagation();
        },
        preventDefault() {
          e.preventDefault();
        },
        nativeEvent: e.nativeEvent,
        reactEvent: e
      };
      if (context && (0, _util.isFunction)(context.onChange)) {
        context.onChange(eventParams);
      } else {
        setCheckedState(e.target.checked);
        (0, _util.isFunction)(onChange) && onChange(eventParams);
      }
    },
    "aria-checked": checkedState,
    disabled: disableState,
    name: nameState
  }, (0, _util.omit)(restProps, ['error']))), /*#__PURE__*/_react.default.createElement("span", {
    className: cssScope('__wallpaper')
  }), /*#__PURE__*/_react.default.createElement("span", {
    className: cssScope('__checked-ink')
  })), children ? /*#__PURE__*/_react.default.createElement(_text.Text, {
    variant: "plain",
    className: (0, _clsx.default)(cssScope('__label-content'), disableState && cssScope('__label-content-disabled')),
    as: "span",
    lineClamp: isButton ? 1 : undefined,
    sx: sxLabelContent
  }, children) : null);
});
var _default = exports.default = Radio;
