// @universe-design/react/lib/form/FormItemUI.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FormItemUI = void 0;
var _u10wqw = _interopRequireDefault(require("./../.stylex/u10wqw00"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _InfoOutlined = _interopRequireDefault(require("@universe-design/icons-react/InfoOutlined"));
var _clsx = _interopRequireDefault(require("clsx"));
var _react = _interopRequireDefault(require("react"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _col = require("../col");
var _row = require("../row");
var _hooks = require("../config-provider/hooks");
var _tooltip = require("../tooltip");
var _util = require("../util");
var _FormUIContext = require("./FormUIContext");
var _forwardRef = require("../util/forwardRef");
var _text = require("../text");
require("./style");
//.u1e8zt0z{min-width:min-content;}
/**
 * @public
 */
/**
 * @public
 */
/**
 * @public
 * @docgen
 */
const FormItemUI = exports.FormItemUI = (0, _forwardRef.forwardHTMLRef)((props, ref) => {
  var _ref, _ref2, _ref3, _ref4;
  const {
    label,
    className,
    style,
    children,
    labelAlign,
    labelCol,
    required = false,
    wrapperCol,
    help,
    hidden,
    validateStatus,
    tooltip,
    extra,
    htmlFor,
    labelInfoIconRenderer,
    errorDisplay: errorDisplayProp,
    ...restProps
  } = props;
  const formCtxLabelAlign = (0, _FormUIContext.useFormUIContextSelector)(ctx => ctx.labelAlign);
  const formCtxLabelCol = (0, _FormUIContext.useFormUIContextSelector)(ctx => ctx.labelCol);
  const formCtxWrapperCol = (0, _FormUIContext.useFormUIContextSelector)(ctx => ctx.wrapperCol);
  const formCtxErrorDisplay = (0, _FormUIContext.useFormUIContextSelector)(ctx => ctx.errorDisplay);
  const mergedLabelAlign = (_ref = labelAlign !== null && labelAlign !== void 0 ? labelAlign : formCtxLabelAlign) !== null && _ref !== void 0 ? _ref : 'right';
  const mergedLabelCol = (_ref2 = labelCol !== null && labelCol !== void 0 ? labelCol : formCtxLabelCol) !== null && _ref2 !== void 0 ? _ref2 : {};
  const mergedWrapperCol = (_ref3 = wrapperCol !== null && wrapperCol !== void 0 ? wrapperCol : formCtxWrapperCol) !== null && _ref3 !== void 0 ? _ref3 : {};
  const errorDisplay = (_ref4 = errorDisplayProp !== null && errorDisplayProp !== void 0 ? errorDisplayProp : formCtxErrorDisplay) !== null && _ref4 !== void 0 ? _ref4 : 'locally';
  const itemCssScope = (0, _hooks.useCssScope)('form__item');
  const itemLabelCssScope = (0, _hooks.useCssScope)('form__item__label');
  const itemControlCssScope = (0, _hooks.useCssScope)('form__item__control');
  const hasExplain = (0, _util.isNonNullable)(help) && errorDisplay !== 'none';
  /**
   * ================ Tooltip Content Node ================
   */
  const {
    tooltipProps,
    labelInfoIconNode
  } = (() => {
    let _tooltipProps = {};
    if (tooltip && typeof tooltip === 'object' && ! /*#__PURE__*/_react.default.isValidElement(tooltip)) {
      _tooltipProps = tooltip;
    } else {
      _tooltipProps = {
        title: tooltip
      };
    }
    const {
      icon = /*#__PURE__*/_react.default.createElement(_InfoOutlined.default, null),
      ...restTooltipProps
    } = _tooltipProps;
    const _labelInfoIconNode = /*#__PURE__*/_react.default.createElement(_text.Text, {
      as: "div",
      className: itemCssScope('__tooltip')
    }, icon);
    return {
      tooltipProps: restTooltipProps,
      labelInfoIconNode: _labelInfoIconNode
    };
  })();
  const tooltipNode = (() => {
    if ((0, _util.isFunction)(labelInfoIconRenderer)) {
      return labelInfoIconRenderer(labelInfoIconNode);
    }
    if (tooltip) {
      return /*#__PURE__*/_react.default.createElement(_tooltip.Tooltip, tooltipProps, labelInfoIconNode);
    }
    return null;
  })();

  /**
   * ================ Extra Node ================
   */
  const extraNode = extra ? /*#__PURE__*/_react.default.createElement(_text.Text, {
    className: itemControlCssScope('__extra')
  }, extra) : null;

  /**
   * ================ Error Or Help Node ================
   */
  const errorOrHelpNode = hasExplain ? /*#__PURE__*/_react.default.createElement(_text.Text, {
    className: (0, _clsx.default)(itemControlCssScope('__explain'), validateStatus === 'error' && itemControlCssScope('__explain-error'))
  }, help) : null;
  return /*#__PURE__*/_react.default.createElement(_row.Row, (0, _extends2.default)({}, restProps, {
    className: (0, _clsx.default)(itemCssScope(), hasExplain && itemCssScope('-with-help'), hidden && itemCssScope('-hidden'), itemCssScope(`-error-${errorDisplay}`), className),
    style: style,
    ref: ref
  }), (0, _util.isNonNullable)(label) ? /*#__PURE__*/_react.default.createElement(_col.Col, (0, _extends2.default)({}, mergedLabelCol, {
    className: (0, _clsx.default)(itemLabelCssScope(), mergedLabelAlign === 'left' && itemLabelCssScope('-left'), mergedLabelCol.className)
  }), /*#__PURE__*/_react.default.createElement(_text.Text, {
    as: "label",
    htmlFor: htmlFor,
    title: typeof label === 'string' ? label : undefined,
    lineClamp: 2,
    sx: [{
      "props": {
        "Bf4jedk": _u10wqw.default
      }
    }]
  }, label), required && /*#__PURE__*/_react.default.createElement(_text.Text, {
    className: itemCssScope('__required-mark')
  }, "*"), tooltipNode) : null, /*#__PURE__*/_react.default.createElement(_col.Col, (0, _extends2.default)({}, mergedWrapperCol, {
    className: (0, _clsx.default)(itemControlCssScope(), mergedWrapperCol.className)
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: itemControlCssScope('__input')
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: itemControlCssScope('__input__content')
  }, children)), errorOrHelpNode, extraNode));
});
if (process.env.NODE_ENV !== "production") {
  FormItemUI.displayName = 'FormItemUI';
}
