// @universe-design/react/lib/form/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Form = void 0;
Object.defineProperty(exports, "FormContextProvider", {
  enumerable: true,
  get: function () {
    return _reactMiniForm.FormContextProvider;
  }
});
exports.FormItem = void 0;
exports.FormList = FormList;
Object.defineProperty(exports, "FormSelector", {
  enumerable: true,
  get: function () {
    return _reactMiniForm.FormSelector;
  }
});
Object.defineProperty(exports, "FormSelectors", {
  enumerable: true,
  get: function () {
    return _reactMiniForm.FormSelectors;
  }
});
Object.defineProperty(exports, "useField", {
  enumerable: true,
  get: function () {
    return _reactMiniForm.useField;
  }
});
Object.defineProperty(exports, "useFieldContext", {
  enumerable: true,
  get: function () {
    return _reactMiniForm.useFieldContext;
  }
});
Object.defineProperty(exports, "useForm", {
  enumerable: true,
  get: function () {
    return _reactMiniForm.useForm;
  }
});
Object.defineProperty(exports, "useFormContext", {
  enumerable: true,
  get: function () {
    return _reactMiniForm.useFormContext;
  }
});
Object.defineProperty(exports, "useFormContextSelector", {
  enumerable: true,
  get: function () {
    return _reactMiniForm.useFormContextSelector;
  }
});
Object.defineProperty(exports, "useFormList", {
  enumerable: true,
  get: function () {
    return _reactMiniForm.useFormList;
  }
});
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _clsx = _interopRequireDefault(require("clsx"));
var _react = _interopRequireWildcard(require("react"));
var _merge = _interopRequireDefault(require("lodash/merge"));
var _get = _interopRequireDefault(require("lodash/get"));
var _scrollIntoViewIfNeeded = _interopRequireDefault(require("scroll-into-view-if-needed"));
var _reactMiniForm = require("../react-mini-form");
var _hooks = require("../config-provider/hooks");
var _util = require("../util");
var _validateUtil = require("./utils/validateUtil");
var _FormUIContext = require("./FormUIContext");
var _FormItemUI = require("./FormItemUI");
var _shared = require("../react-mini-form/shared");
var _configProvider = require("../config-provider");
var _reactUtilities = require("@universe-design/react-utilities");
var _FormDescendantContext = require("./FormDescendantContext");
require("./style");
/* eslint-disable react/destructuring-assignment */

/**
 * @public
 */

/**
 * @public
 * @docgen
 */

/**
 * @docgen
 * @public
 * 表单字段组件`Form.Item`，用于数据双向绑定、校验、布局等。
 */

/**
 * @docgen
 * @public
 * @typeParam Values - Values of this form
 */

/**
 * @public
 */

/**
 * @docgen
 * @public
 * @typeParam FieldValue - FieldValue of this form list
 */

const BaseForm = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  const size = (0, _hooks.useComponentSize)();
  const {
    layout = 'vertical',
    className,
    requiredMark = true,
    size: sizeProp = size,
    labelAlign,
    labelCol,
    wrapperCol,
    onFinish,
    validateMessages,
    errorDisplay,
    component = 'form',
    children,
    shouldScrollToFirstErrorField = true,
    onSubmitError,
    ...restProps
  } = props;

  /**
   * ================ ClassName ================
   */
  const formCssScope = (0, _hooks.useCssScope)('form');
  const formClassName = (0, _clsx.default)(formCssScope(), formCssScope(`-${layout}`), !requiredMark && formCssScope('-hide-required-mark'), formCssScope(`-${(0, _util.getDefaultSizeAbbr)(sizeProp, 'medium')}`), className);

  /**
   * ================ Logic context ================
   */
  const formCtx = (0, _reactMiniForm.useForm)({
    ...restProps,
    onSubmit: onFinish,
    onSubmitError: errors => {
      (0, _util.isFunction)(onSubmitError) && onSubmitError(errors);
      // eslint-disable-next-line @typescript-eslint/no-use-before-define
      scrollToFirstError();
    }
  });
  const {
    submit
  } = formCtx;

  /**
   * ================ Descendants ================
   */
  const descendants = (0, _FormDescendantContext.useFormDescendants)();

  /**
   * ================ Methods ================
   */

  // ---------------- Handle submit ----------------
  const handleSubmit = _react.default.useCallback(e => {
    e.preventDefault();
    e.stopPropagation();
    submit();
  }, [submit]);
  let childNode = children;
  if (component) {
    childNode = /*#__PURE__*/_react.default.createElement("form", {
      onSubmit: handleSubmit,
      className: formClassName
    }, childNode);
  }

  // ---------------- scrollToFirstError ----------------
  const scrollToFirstError = _react.default.useCallback(option => {
    if (shouldScrollToFirstErrorField && formCtx.isDirty) {
      var _descendants$values$f;
      const node = (_descendants$values$f = descendants.values().find(o => o.id && (0, _get.default)(formCtx.errors, (0, _shared.unmarshalNamePath)(o.id)))) === null || _descendants$values$f === void 0 ? void 0 : _descendants$values$f.node;
      if (node) {
        const mergedOption = option !== null && option !== void 0 ? option : shouldScrollToFirstErrorField === true ? {
          block: 'nearest'
        } : shouldScrollToFirstErrorField;
        (0, _scrollIntoViewIfNeeded.default)(node, {
          scrollMode: 'if-needed',
          ...(mergedOption || {})
        });
      }
    }
  }, [descendants, formCtx.errors, formCtx.isDirty, shouldScrollToFirstErrorField]);

  /**
   * ================ Ref ================
   */
  (0, _react.useImperativeHandle)(ref, () => Object.assign(Object.create(Object.getPrototypeOf(formCtx), Object.getOwnPropertyDescriptors(formCtx)), {
    scrollToFirstError
  }));
  return /*#__PURE__*/_react.default.createElement(_FormDescendantContext.FormDescendantsProvider, {
    value: descendants
  }, /*#__PURE__*/_react.default.createElement(_configProvider.ConfigProvider, {
    componentSize: sizeProp
  }, /*#__PURE__*/_react.default.createElement(_FormUIContext.FormUIContextProvider, {
    value: {
      requiredMark,
      labelAlign,
      labelCol,
      wrapperCol,
      validateMessages,
      errorDisplay
    }
  }, /*#__PURE__*/_react.default.createElement(_reactMiniForm.FormContextProvider, {
    value: formCtx
  }, childNode))));
});
function useValidators(props) {
  const {
    label,
    rules,
    validate: validateProp
  } = props;
  const formCtxValidateMessages = (0, _FormUIContext.useFormUIContextSelector)(ctx => ctx.validateMessages);
  const locale = (0, _hooks.useLocaleReceiver)('form');
  const variables = (0, _react.useMemo)(() => {
    const _variables = {};
    if (typeof label === 'string') {
      _variables.label = label;
    } else {
      _variables.label = '';
    }
    return _variables;
  }, [label]);
  const validate = _react.default.useMemo(() => {
    const validators = [];
    if (rules) {
      validators.push(...(0, _shared.asArray)(rules).map(rule => {
        const validator = {
          validateTrigger: rule.validateTrigger,
          validate: async (value, _name) => {
            const namePathString = (0, _reactMiniForm.marshalNamePath)(_name);
            const errorMessages = await (0, _validateUtil.validateRule)(namePathString, value, rule, {
              validateMessages: (0, _merge.default)({}, formCtxValidateMessages, locale.defaultValidateMessage)
            }, variables);
            if (errorMessages.length > 0) {
              return errorMessages[0];
            }
          }
        };
        return validator;
      }));
    }
    if (validateProp) {
      validators.push(...(0, _shared.asArray)(validateProp));
    }
    return validators;
  }, [locale.defaultValidateMessage, formCtxValidateMessages, rules, validateProp, variables]);
  return validate;
}
function _FormItem(props) {
  const {
    noStyle = false,
    label,
    className,
    style,
    name,
    children,
    labelAlign,
    labelCol,
    rules,
    // eslint-disable-next-line no-nested-ternary
    required = rules ? Array.isArray(rules) ? rules.some(r => r.required) : rules.required : undefined,
    wrapperCol,
    help,
    validateStatus,
    tooltip,
    extra,
    hidden,
    htmlFor,
    validate: validateProp,
    labelInfoIconRenderer,
    errorDisplay,
    ...restProps
  } = props;
  const {
    error
  } = (0, _reactMiniForm.useFieldContext)({
    name
  });
  const mergedValidateStatus = validateStatus !== null && validateStatus !== void 0 ? validateStatus : error ? 'error' : undefined;
  process.env.NODE_ENV !== "production" ? (0, _util.warning)((0, _util.isNonNullable)(name), (0, _util.devWarning)('FormItem', 'The `name` prop is undefined, thus this field will not be registered to the form')) : void 0;
  const validate = useValidators({
    label,
    validate: validateProp,
    rules
  });

  /**
   * Descendant
   */
  const id = (0, _util.isNonNullable)(name) ? (0, _reactMiniForm.marshalNamePath)(name) : undefined;
  const {
    register
  } = (0, _FormDescendantContext.useFormDescendant)(_react.default.useMemo(() => ({
    id
  }), [id]));

  /**
   * ================ Input Content Node ================
   *
   * child type: funcChild | reactNodeChild
   *
   * noStyle: true | false
   *
   * - 如果 noStyle === false，则直接用 FormItemUI 的 ref 来 register
   * - 如果 noStyle === true，则需要判断 children 类型
   *  - 如果 children 是 ReactElement 则直接套住 FindDOMNode
   *  - 如果 children 是 function 则先交给 Field 处理，然后再套住 FindDOMNode
   */
  let inputContentNode = children;
  // 判断是否是一个 field 组件，如果没有 name 的话则认为只是一个 UI submit 而不是 field
  if (props.name !== undefined && (/*#__PURE__*/(0, _react.isValidElement)(children) || (0, _util.isFunction)(children))) {
    inputContentNode = /*#__PURE__*/_react.default.createElement(_reactMiniForm.Field, (0, _extends2.default)({
      name: props.name,
      validate: validate,
      validateStatus: mergedValidateStatus
    }, restProps), (() => {
      if (!noStyle) {
        return children;
      }
      if ((0, _util.isFunction)(children)) {
        return fieldProps => /*#__PURE__*/_react.default.createElement(_reactUtilities.FindDOMNode, {
          domOnly: true,
          childRef: register
        }, children(fieldProps));
      }
      return /*#__PURE__*/_react.default.createElement(_reactUtilities.FindDOMNode, (0, _extends2.default)({}, children.props, {
        domOnly: true,
        childRef: register
      }), children);
    })());
  }
  if (noStyle) {
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, inputContentNode);
  }
  return /*#__PURE__*/_react.default.createElement(_FormItemUI.FormItemUI, {
    label,
    className,
    style,
    children,
    labelAlign,
    labelCol,
    required,
    wrapperCol,
    help: help !== null && help !== void 0 ? help : error,
    hidden,
    validateStatus: mergedValidateStatus,
    tooltip,
    extra,
    htmlFor,
    labelInfoIconRenderer,
    errorDisplay,
    ref: register
  }, inputContentNode);
}
function FormList(props) {
  const {
    validate,
    rules,
    ...others
  } = props;
  const validators = useValidators({
    rules,
    validate
  });
  return /*#__PURE__*/_react.default.createElement(_reactMiniForm.FormList, (0, _extends2.default)({
    validate: validators
  }, others));
}
const Form = exports.Form = BaseForm;
const FormItem = exports.FormItem = /*#__PURE__*/(0, _react.memo)(_FormItem, _reactMiniForm.FieldPropsAreEqual);
Form.Item = FormItem;
Form.ItemUI = _FormItemUI.FormItemUI;
if (process.env.NODE_ENV !== "production") {
  FormItem.displayName = 'FormItem';
}
