// @universe-design/react/es/timeline/TimelineItem.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React from 'react';
import clsx from 'clsx';
import TimeFilled from '@universe-design/icons-react/TimeFilled';
import TimeOutlined from '@universe-design/icons-react/TimeOutlined';
import CloseFilled from '@universe-design/icons-react/CloseFilled';
import WarningFilled from '@universe-design/icons-react/WarningFilled';
import YesFilled from '@universe-design/icons-react/YesFilled';
import YesOutlined from '@universe-design/icons-react/YesOutlined';
import { useCssScope } from '../config-provider/hooks';
import stylex from '@universe-design/stylex';
import { cssTypographyBody0, cssTypographyBody2 } from '../stylex';
import { SpinLoadingSvg } from '../_assets/svgs';
const useTimelineItemCssScope = () => useCssScope('timeline-item');
const iconStatusMap = {
  success: /*#__PURE__*/React.createElement(YesFilled, null),
  warning: /*#__PURE__*/React.createElement(WarningFilled, null),
  error: /*#__PURE__*/React.createElement(CloseFilled, null),
  done: /*#__PURE__*/React.createElement(YesOutlined, null),
  current: /*#__PURE__*/React.createElement(TimeFilled, null),
  notarrived: /*#__PURE__*/React.createElement(TimeOutlined, null),
  loading: /*#__PURE__*/React.createElement(SpinLoadingSvg, null)
};
const TimelineItem = props => {
  const {
    position,
    labelPosition = 'relative',
    className,
    title,
    label,
    description,
    status,
    dotType = 'dot',
    dot = dotType === 'icon' && status && iconStatusMap[status],
    ...reset
  } = props;
  const itemCssScope = useTimelineItemCssScope();
  const labelEl = /*#__PURE__*/React.createElement("div", {
    className: clsx(itemCssScope('__label'), stylex(cssTypographyBody2))
  }, label);
  const same = labelPosition === 'same';
  return /*#__PURE__*/React.createElement("li", _extends({}, reset, {
    className: clsx(itemCssScope(), className, {
      [itemCssScope('--custom-dot')]: dot,
      [itemCssScope(`--position-${position}`)]: position,
      [itemCssScope(`--status-${status}`)]: status
    })
  }), /*#__PURE__*/React.createElement("div", {
    className: itemCssScope('__oppsite-content')
  }, !same && label && labelEl), /*#__PURE__*/React.createElement("div", {
    className: itemCssScope('__seperator')
  }, dot ? /*#__PURE__*/React.createElement("div", {
    className: itemCssScope('__dot--custom')
  }, dot) : /*#__PURE__*/React.createElement("div", {
    className: itemCssScope('__dot')
  }), /*#__PURE__*/React.createElement("div", {
    className: itemCssScope('__tail')
  })), /*#__PURE__*/React.createElement("div", {
    className: itemCssScope('__content')
  }, title && /*#__PURE__*/React.createElement("div", {
    className: clsx(itemCssScope('__title'), stylex(cssTypographyBody0))
  }, title), same && label && labelEl, description && /*#__PURE__*/React.createElement("div", {
    className: clsx(itemCssScope('__des'), stylex(cssTypographyBody2))
  }, description)));
};
if (process.env.NODE_ENV !== "production") {
  TimelineItem.displayName = 'Timeline.Item';
}
export default TimelineItem;
