// @universe-design/react/es/timeline/Timeline.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React from 'react';
import clsx from 'clsx';
import { useCssScope } from '../config-provider/hooks';
import TimelineItem from './TimelineItem';
import { getDefaultSizeAbbr } from '../util';
import { Loading } from '../loading';
import './style';
const useTimelineCssScope = () => useCssScope('timeline');
const useTimelineItemCssScope = () => useCssScope('timeline-item');
const Timeline = props => {
  const {
    mode = 'left',
    reverse,
    hangingTail,
    children,
    pending,
    pendingDot,
    space = 'medium',
    style,
    className,
    labelPosition = 'relative',
    ...restProps
  } = props;
  const timelineItems = React.Children.toArray(children);
  const validTimelineItems = timelineItems.filter(item => /*#__PURE__*/React.isValidElement(item));
  const hasLabel = validTimelineItems.some(item => {
    var _item$props;
    return Boolean(item === null || item === void 0 ? void 0 : (_item$props = item.props) === null || _item$props === void 0 ? void 0 : _item$props.label);
  });
  const cssScope = useTimelineCssScope();
  const itemCssScope = useTimelineItemCssScope();
  const items = React.Children.map(validTimelineItems, (item, index) => /*#__PURE__*/React.cloneElement(item, {
    className: clsx(mode === 'alternate' ? itemCssScope(`--mode-alt-${index % 2 === 0 ? 'left' : 'right'}`) : itemCssScope(`--mode-${mode}`), {
      [itemCssScope('-last')]: pending && !reverse && index === validTimelineItems.length - 1
    }, item.props.className),
    key: index,
    labelPosition
  }));
  if (pending) {
    const pendingNode = /*#__PURE__*/React.createElement(TimelineItem, {
      status: "current",
      className: clsx(mode === 'alternate' ? itemCssScope('--mode-alt-left') : itemCssScope(`--mode-${mode}`), itemCssScope('--pending')),
      key: "pending",
      dot: pendingDot || /*#__PURE__*/React.createElement(Loading, {
        className: itemCssScope('--loading')
      }),
      title: typeof pending === 'boolean' ? null : pending,
      labelPosition: labelPosition
    });
    items.push(pendingNode);
  }
  if (reverse) {
    items.reverse();
  }
  return /*#__PURE__*/React.createElement("ul", _extends({
    style: style
  }, restProps, {
    className: clsx(cssScope(), {
      [cssScope('--no-label')]: !hasLabel,
      [cssScope('__hanging-tail')]: hangingTail,
      [cssScope('-pending')]: pending
    }, cssScope(`--space-${getDefaultSizeAbbr(space, 'medium')}`), className)
  }), items);
};
Timeline.Item = TimelineItem;
if (process.env.NODE_ENV !== "production") {
  Timeline.displayName = 'Timeline';
}
export default Timeline;
