// @universe-design/react-utilities/lib-esm/domain-events/dom.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
// Domains can only emit their own events, but they can subscribe to their sub-domains events.
import { requestIdleCallback } from '../dom/requestIdleCallback';

/**
 * This is an abstract class of a domain event.
 *
 * @alpha
 */
export class DomainEvent {
  constructor(_ref) {
    let {
      detail,
      bubbles = true,
      cancelable = true
    } = _ref;
    _defineProperty(this, "type", void 0);
    _defineProperty(this, "detail", void 0);
    _defineProperty(this, "timeStamp", Date.now());
    // eventId = '';
    _defineProperty(this, "_cancelable", void 0);
    _defineProperty(this, "_bubbles", void 0);
    _defineProperty(this, "_defaultPrevented", false);
    _defineProperty(this, "_propagationStopped", false);
    _defineProperty(this, "preventDefault", () => {
      if (!this._cancelable) {
        return;
      }
      this._defaultPrevented = true;
    });
    _defineProperty(this, "isDefaultPrevented", () => this._defaultPrevented);
    _defineProperty(this, "stopPropagation", () => {
      this._propagationStopped = true;
    });
    _defineProperty(this, "isPropagationStopped", () => !this._bubbles || this._propagationStopped);
    this._cancelable = cancelable;
    this._bubbles = bubbles;
    this.detail = detail;
  }
}

/**
 * Domain event handler callback function type
 *
 * @alpha
 */

/**
 * `DomainEventTarget` class is similar with DOM EventTarget API which contains a callbacks map and some functions to add listeners and invoke listeners.
 *
 * @alpha
 */
export class DomainEventTarget {
  constructor() {
    _defineProperty(this, "_callbacksMap", new WeakMap());
    _defineProperty(this, "parentEventTarget", null);
    _defineProperty(this, "addEventListener", (type, cb) => {
      const _cbSet = this._callbacksMap.get(type);
      const cbSet = _cbSet !== null && _cbSet !== void 0 ? _cbSet : new Set();
      if (!_cbSet) {
        this._callbacksMap.set(type, cbSet);
      }
      cbSet.add(cb);
      return () => {
        cbSet.delete(cb);
      };
    });
    _defineProperty(this, "invokeEventListeners", event => {
      var _this$_callbacksMap$g;
      const cbSet = (_this$_callbacksMap$g = this._callbacksMap.get(event.constructor)) !== null && _this$_callbacksMap$g !== void 0 ? _this$_callbacksMap$g : new Set();
      for (const cb of cbSet) {
        try {
          cb(event);
        } catch (err) {
          // The logic below might not be correct, mainly trying to simulate error handling in DOM events.
          requestIdleCallback(() => {
            throw err;
          });
        }
      }
    });
    _defineProperty(this, "dispatchEvent", event => {
      let currentEventTarget = this;
      while (currentEventTarget) {
        currentEventTarget.invokeEventListeners(event);
        if (event.isPropagationStopped()) {
          break;
        }
        currentEventTarget = currentEventTarget.parentEventTarget;
      }
      return event.isDefaultPrevented();
    });
  }
}
