// @universe-design/react/es/date-picker/panels/selector/MonthYearSelector.js
import React from 'react';
import clsx from 'clsx';
import stylex from '@universe-design/stylex';
import DownBoldOutlined from '@universe-design/icons-react/DownBoldOutlined';
import { useCssScope, useLocale } from '../../../config-provider/hooks';
import { cssTypographyHeadline } from '../../../stylex';
import { Button } from '../../../button';
import { getYearRange } from '../base/utils';
export const MonthYearSelector = props => {
  const {
    className,
    mode,
    collapsed,
    pickerValue,
    hiddenCollapse,
    onYearClick,
    onMonthClick,
    onCollapsed
  } = props;
  const {
    locale: localeName,
    components: {
      datePicker: locale
    }
  } = useLocale();
  const cssScope = useCssScope('picker');
  const headerPrefixCls = cssScope('-panel-header');
  const wrapYear = children => /*#__PURE__*/React.createElement("span", {
    key: "year",
    className: `${headerPrefixCls}-btn`,
    onClick: onYearClick
  }, children);
  const timeNodes = locale.monthYearSelector.f({
    year: yearFormat => wrapYear(pickerValue.clone().locale(localeName).format(yearFormat.toString())),
    month: monthFormat => /*#__PURE__*/React.createElement("span", {
      key: "month",
      onClick: onMonthClick,
      className: `${headerPrefixCls}-btn`
    }, pickerValue.clone().locale(localeName).format(monthFormat.toString()))
  });
  return /*#__PURE__*/React.createElement("div", {
    className: clsx(`${headerPrefixCls}-content`, className, stylex(cssTypographyHeadline))
  }, mode === 'year' ? wrapYear(getYearRange(pickerValue)) : timeNodes, !hiddenCollapse && /*#__PURE__*/React.createElement(Button, {
    type: "icon",
    className: clsx(`${headerPrefixCls}-icon`, `${headerPrefixCls}-collapse`),
    onClick: e => {
      e.stopPropagation();
      e.preventDefault();
      onCollapsed === null || onCollapsed === void 0 ? void 0 : onCollapsed(!collapsed);
    },
    size: "extra-small",
    icon: /*#__PURE__*/React.createElement(DownBoldOutlined, {
      className: clsx({
        [`${headerPrefixCls}-rotate`]: true,
        [`${headerPrefixCls}-rotate180`]: !collapsed
      })
    })
  }));
};
