// @universe-design/react-utilities/lib-esm/hooks/useCloneElementRef.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
/* eslint-disable react/no-find-dom-node */
/* eslint-disable react/prefer-stateless-function */
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { isNonNullable } from '../util';
import { isElement, isHTMLElement, isSVGElement } from '../dom';
import { supportRef, updateRef } from './composeRefs';
const getChildrenRef = child => child === null || child === void 0 ? void 0 : child.ref;

/**
 * 用于获取子组件的 dom ref
 *
 * @public
 */
export class FindDOMNode extends React.Component {
  constructor() {
    super(...arguments);
    _defineProperty(this, "node", void 0);
    _defineProperty(this, "hasRefFuncBeenCalled", false);
    _defineProperty(this, "isRefSupported", true);
    _defineProperty(this, "updateChildRef", node => {
      if (this.node === node) {
        return;
      }
      if (this.domOnly && node !== null && !isHTMLElement(node) && !isSVGElement(node)) {
        return;
      }
      updateRef(this.props.childRef, node);
      this.node = node;
    });
    _defineProperty(this, "updateRefWhenLifecycle", () => {
      // 如果不是 domOnly，那就不需要额外操作 ref 了
      if (!this.domOnly) {
        return;
      }

      // 如果是 domOnly，并且不支持 ref 那就需要在 lifecycle 做更新
      const {
        children
      } = this.props;
      if (!supportRef(children)) {
        this.updateChildRef(ReactDOM.findDOMNode(this));
        return;
      }

      // 如果是 domOnly，并且支持 ref，需要兜底一种特殊情况：组件支持 ref 但是内部没有实际调用 ref
      if (!this.hasRefFuncBeenCalled) {
        this.updateChildRef(ReactDOM.findDOMNode(this));
        this.hasRefFuncBeenCalled = false;
      }
    });
    _defineProperty(this, "updateRefWhenSupportRef", node => {
      const {
        children
      } = this.props;
      const origRef = getChildrenRef(children);
      if (origRef) {
        updateRef(origRef, node);
      }
      if (node === null) {
        this.updateChildRef(null);
        this.hasRefFuncBeenCalled = false;
        return;
      }
      this.hasRefFuncBeenCalled = true;
      if (this.domOnly && node && !isElement(node)) {
        this.updateChildRef(ReactDOM.findDOMNode(this));
        return;
      }
      this.updateChildRef(node);
    });
  }
  get domOnly() {
    const {
      domOnly = true
    } = this.props;
    return domOnly;
  }
  componentDidMount() {
    this.updateRefWhenLifecycle();
  }
  componentDidUpdate() {
    this.updateRefWhenLifecycle();
  }
  componentWillUnmount() {
    if (!this.isRefSupported) {
      updateRef(this.props.childRef, null);
      this.hasRefFuncBeenCalled = false;
    }
  }
  render() {
    this.isRefSupported = false;
    const {
      children: childrenProp,
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      childRef,
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      domOnly,
      ...restProps
    } = this.props;
    if (! /*#__PURE__*/React.isValidElement(childrenProp)) {
      return childrenProp;
    }
    // forward props
    const children = /*#__PURE__*/React.cloneElement(childrenProp, restProps);
    if (!supportRef(children)) {
      return children;
    }
    this.isRefSupported = true;
    return /*#__PURE__*/React.cloneElement(children, {
      ref: this.updateRefWhenSupportRef
    });
  }
}

/**
 * 给 child node clone ref
 *
 * @public
 */
export const useCloneElementRef = function (_ref) {
  let {
    children,
    ref
  } = _ref;
  let {
    skipped = false,
    domOnly
  } = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  if (skipped || !isNonNullable(children)) {
    return children;
  }
  return /*#__PURE__*/React.createElement(FindDOMNode, {
    domOnly: domOnly,
    childRef: ref
  }, children);
};
