// @byted-service/bytedtrace/lib/metrics/metrics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ByteTraceMetrics = void 0;
const tslib_1 = require("tslib");
const constants = tslib_1.__importStar(require("../span/consts"));
const utils_1 = require("../utils");
const metrics_1 = require("@byted-nodex/metrics");
const SERVER_RECEIVE_BYTES_NAME = 'span.server.receive.bytes';
const SERVER_SEND_BYTES_NAME = 'span.server.send.bytes';
const CLIENT_RECEIVE_BYTES_NAME = 'span.client.receive.bytes';
const CLIENT_SEND_BYTES_NAME = 'span.client.send.bytes';
class ByteTraceMetrics {
    constructor(options, globalTags = {}) {
        this.globalTags = Object.assign({}, globalTags);
        this.client = new metrics_1.CompatClient(Object.assign({ metricOptions: { tenant: 'apm.rpc' } }, options));
        this.clientTimerMetric = this.client.newMetric(constants.CLIENT_SPAN_METRICS_TIMER_NAME, {
            compatTimer: false,
        });
        this.serverTimerMetric = this.client.newMetric(constants.SERVICE_SPAN_METRICS_TIMER_NAME, {
            compatTimer: false,
        });
        this.serverRecvSizeRateCounter = this.client.newMetric(SERVER_RECEIVE_BYTES_NAME);
        this.serverSendSizeRateCounter = this.client.newMetric(SERVER_SEND_BYTES_NAME);
        this.clientRecvSizeRateCounter = this.client.newMetric(CLIENT_RECEIVE_BYTES_NAME);
        this.clientSendSizeRateCounter = this.client.newMetric(CLIENT_SEND_BYTES_NAME);
        this.serverRateCounter = this.client.newMetric(constants.SERVICE_SPAN_METRICS_COUNTER_NAME);
        this.clientRateCounter = this.client.newMetric(constants.CLIENT_SPAN_METRICS_COUNTER_NAME);
        for (const key of ['emitCounter', 'emitStore', 'emitMeter']) {
            this[key] = function (name, value, prefix, tags, tagKeys) {
                tags = (0, utils_1.transPropertyToTarget)(tags, this.globalTags);
                this.client[key](name, value, prefix, tags, tagKeys);
            };
        }
    }
    emitTimer(name, value, prefix, tags, tagKeys) {
        tags = (0, utils_1.transPropertyToTarget)(tags, this.globalTags);
        if (name === constants.CLIENT_SPAN_METRICS_TIMER_NAME) {
            this.clientTimerMetric
                .withTags(tags)
                .emit((0, metrics_1.withSuffix)('').observe(value));
        }
        else if (name === constants.SERVICE_SPAN_METRICS_TIMER_NAME) {
            this.serverTimerMetric
                .withTags(tags)
                .emit((0, metrics_1.withSuffix)('').observe(value));
        }
        else {
            this._emitTimer(name, value, prefix, tags, tagKeys);
        }
    }
    emitRateCounter(name, value, prefix, tags, tagKeys) {
        tags = (0, utils_1.transPropertyToTarget)(tags, this.globalTags);
        switch (name) {
            // send/receive size
            case SERVER_RECEIVE_BYTES_NAME:
                this.serverRecvSizeRateCounter.emit((0, metrics_1.withSuffix)('').incr(value));
                break;
            case SERVER_SEND_BYTES_NAME:
                this.serverSendSizeRateCounter.emit((0, metrics_1.withSuffix)('').incr(value));
                break;
            case CLIENT_RECEIVE_BYTES_NAME:
                this.clientRecvSizeRateCounter.emit((0, metrics_1.withSuffix)('').incr(value));
                break;
            case CLIENT_SEND_BYTES_NAME:
                this.clientSendSizeRateCounter.emit((0, metrics_1.withSuffix)('').incr(value));
                break;
            // rate
            case constants.SERVICE_SPAN_METRICS_COUNTER_NAME:
                this.serverRateCounter
                    .withTags(tags)
                    .emit((0, metrics_1.withSuffix)('').incr(value));
                break;
            case constants.CLIENT_SPAN_METRICS_COUNTER_NAME:
                this.clientRateCounter
                    .withTags(tags)
                    .emit((0, metrics_1.withSuffix)('').incr(value));
                break;
            default:
                this._emitRateCounter(name, value, prefix, tags, tagKeys);
        }
    }
    _emitTimer(name, value, prefix, tags, tagKeys) {
        this.client.emitTimer(name, value, prefix, tags, tagKeys);
    }
    _emitRateCounter(name, value, prefix, tags, tagKeys) {
        this.client.emitRateCounter(name, value, prefix, tags, tagKeys);
    }
}
exports.ByteTraceMetrics = ByteTraceMetrics;
