// @universe-design/react/es/image-viewer/ImageViewerImage.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.uk73vx1{opacity:0;}
import usfh9mg from "./../.stylex/usfh9mg";
//.u1825aan{color:@N400;}
import u7iz6qb from "./../.stylex/u7iz6qb";
//.u18m8u13{font-size:48px;}
import ugj19bf from "./../.stylex/ugj19bf";
//.uwfqyga{width:120px;}
import u1it5fv0 from "./../.stylex/u1it5fv0";
//.ujr5b71{height:120px;}
import u1b894eo from "./../.stylex/u1b894eo";
//.u1t7pc3r{background:@N100;}
import uhiotwf from "./../.stylex/uhiotwf";
//.u5ogflp{border-top-width:1px;}
import u1cpcpb5 from "./../.stylex/u1cpcpb5";
//.u1hqa2wf{border-right-width:1px;}
import u10qn3go from "./../.stylex/u10qn3go";
//.u1f09k3d{border-bottom-width:1px;}
import uflkg16 from "./../.stylex/uflkg16";
//.uinvdd3{border-left-width:1px;}
import uvn5txf from "./../.stylex/uvn5txf";
//.uzkkow9{border-top-style:solid;}
import us9j13g from "./../.stylex/us9j13g";
//.ucdblym{border-right-style:solid;}
import umt7kug from "./../.stylex/umt7kug";
//.ug706s2{border-bottom-style:solid;}
import uh3lwu4 from "./../.stylex/uh3lwu4";
//.ujik90z{border-left-style:solid;}
import ur61548 from "./../.stylex/ur61548";
//.uyxjys6{border-top-color:@line-border-card;}
import u1n7c7ic from "./../.stylex/u1n7c7ic";
//.u1f4hk3w{border-right-color:@line-border-card;}
import ujkvyav from "./../.stylex/ujkvyav";
//.u1x626fe{border-bottom-color:@line-border-card;}
import u182xfmz from "./../.stylex/u182xfmz";
//.u14wfv0u{border-left-color:@line-border-card;}
import uqbl7tl from "./../.stylex/uqbl7tl";
//.ugvjd9r{border-top-left-radius:@border-radius-xs;}
import urnexrx from "./../.stylex/urnexrx";
//.uns275c{border-top-right-radius:@border-radius-xs;}
import u12hcvx2 from "./../.stylex/u12hcvx2";
//.udbqa1p{border-bottom-right-radius:@border-radius-xs;}
import u1gja64x from "./../.stylex/u1gja64x";
//.u1vdvwtp{border-bottom-left-radius:@border-radius-xs;}
import u162mx8s from "./../.stylex/u162mx8s";
//.u1euv43f{position:absolute;}
import u1u1p7mj from "./../.stylex/u1u1p7mj";
//.u1i1t8d1{top:50%;}
import ufibavv from "./../.stylex/ufibavv";
//.u1ktbn1t{left:50%;}
import u1jmni06 from "./../.stylex/u1jmni06";
//.uoouk8k{transform:translate(-50%, -50%);}
import uevjesf from "./../.stylex/uevjesf";
//.u22iagw{display:flex;}
import ukchsq7 from "./../.stylex/ukchsq7";
//.u4d9j23{justify-content:center;}
import ut7q8es from "./../.stylex/ut7q8es";
//.u122n59{align-items:center;}
import ubyhozq from "./../.stylex/ubyhozq";
//.ujseox{display:none;}
import uh6ryxt from "./../.stylex/uh6ryxt";
//.u1cguypg{pointer-events:auto;}
import u1ksa1gg from "./../.stylex/u1ksa1gg";
//.u1xqy1su{user-select:none;}
import u1443mj0 from "./../.stylex/u1443mj0";
//.u1sdsnyy{touch-action:none;}
import uit42qe from "./../.stylex/uit42qe";
//.u1b6n930{transform-origin:left top;}
import uc4lfdq from "./../.stylex/uc4lfdq";
//.u17vrrt3{-webkit-user-drag:none;}
import u1tngplc from "./../.stylex/u1tngplc";
/* eslint-disable react/no-unknown-property */
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx, makeStylexStyles } from '@universe-design/stylex/react';
import * as React from 'react';
import { forwardRef, useEventCallback, useMergedRefs } from '@universe-design/react-utilities';
import { useImage_unstable } from '../shared/components/Image';
import { useImageViewerContext } from './ImageViewerContext';
import LoadfailFilled from '@universe-design/icons-react/LoadfailFilled';
import stylex from '@universe-design/stylex';
import { resolveSlot } from '../util/slot';
import { composeEventHandlers } from '../util/function';
import { Loading } from '../loading';
import { mat2d, vec2 } from 'gl-matrix';
import { usePointerHandler } from './usePointerHandler';
import { CURRENT_SCALE_OF_ORIGINAL_SIZE, DELTA_Y_SCALE } from './constants';
import { useImageViewerMobileScaling } from './useImageViewerMobileScaling';
import { ImageViewerCloseAction } from './useImageViewerVisible';
const useImageViewerImageStylexStyles = makeStylexStyles(_ref => {
  let {
    prefix
  } = _ref;
  return {
    image: {
      "props": {
        "Bkecrkj": u1ksa1gg,
        "famaaq": u1443mj0,
        "lpbzjs": uit42qe,
        "B76gyc8": uc4lfdq,
        "Ijiw13": u1tngplc
      },
      "alias": prefix
    },
    imageError: {
      "props": {
        "mc9l5x": uh6ryxt
      },
      "alias": `${prefix}-error`
    },
    loading: {
      "props": {
        "qhf8xq": u1u1p7mj,
        "Bhzewxz": ufibavv,
        "oyh7mz": u1jmni06,
        "Bz10aip": uevjesf,
        "mc9l5x": ukchsq7,
        "Brf1p80": ut7q8es,
        "Bt984gj": ubyhozq
      },
      "alias": `${prefix}__loading`
    },
    error: {
      "props": {
        "qhf8xq": u1u1p7mj,
        "Bhzewxz": ufibavv,
        "oyh7mz": u1jmni06,
        "Bz10aip": uevjesf,
        "mc9l5x": ukchsq7,
        "Brf1p80": ut7q8es,
        "Bt984gj": ubyhozq,
        "a9b677": u1it5fv0,
        "Bqenvij": u1b894eo,
        "ayd6f0": uhiotwf,
        "B4j52fo": u1cpcpb5,
        "Bekrc4i": u10qn3go,
        "Bn0qgzm": uflkg16,
        "ibv6hh": uvn5txf,
        "icvyot": us9j13g,
        "vrafjx": umt7kug,
        "oivjwe": uh3lwu4,
        "wvpqe5": ur61548,
        "g2u3we": u1n7c7ic,
        "h3c5rm": ujkvyav,
        "B9xav0g": u182xfmz,
        "zhjwy3": uqbl7tl,
        "Btl43ni": urnexrx,
        "B7oj6ja": u12hcvx2,
        "Bbmb7ep": u1gja64x,
        "Beyfa6y": u162mx8s
      },
      "alias": `${prefix}__error`
    },
    errorIcon: {
      "props": {
        "sj55zd": u7iz6qb,
        "Be2twd7": ugj19bf
      },
      "alias": `${prefix}__error__icon`
    },
    hideImage: {
      "props": {
        "abs64n": usfh9mg
      },
      "alias": `${prefix}__hide`
    }
  };
});
const useImageViewerImage = () => {
  const {
    initTransform,
    setMatrixByPoints,
    scaleByOrigin,
    setDialogVisible
  } = useImageViewerContext();
  const [isImageVisible, setIsImageVisible] = React.useState(false);
  const [isPressing, setIsPressing] = React.useState(false);
  const [prevSrc, setPrevSrc] = React.useState('');

  /**
   * Translation
   *
   * - 按住拖拽（跨端同构，单点，onPointerDown + Move + Up）
   */
  const singleTouchTranslationHandler = usePointerHandler({
    down: () => {
      setIsPressing(true);
    },
    move: (ev, pointerIds) => {
      const lastEvent = pointerIds.get(ev.pointerId);
      if (pointerIds.size === 1 && lastEvent) {
        // 单点触控
        const {
          clientX,
          clientY
        } = lastEvent;
        setMatrixByPoints(point => {
          const translation = mat2d.fromTranslation(mat2d.create(), [ev.clientX - clientX, ev.clientY - clientY]);
          return vec2.transformMat2d(vec2.create(), point, translation);
        });
      }
    },
    up: () => {
      setIsPressing(false);
    },
    tap: event => {
      setDialogVisible(false, {
        visible: false,
        event,
        action: ImageViewerCloseAction.Image
      });
    }
  });

  /**
   * Translation
   *
   * - 触控板滚动（onWheel）
   * - 鼠标滚轮（onWheel）
   */
  React.useEffect(() => {
    const handleWheel = event => {
      const {
        metaKey,
        ctrlKey
      } = event;
      if (metaKey || ctrlKey) {
        return;
      }
      setMatrixByPoints(point => {
        const translation = mat2d.fromTranslation(mat2d.create(), [-event.deltaX, -event.deltaY]);
        return vec2.transformMat2d(vec2.create(), point, translation);
      });
    };
    document.addEventListener('wheel', handleWheel);
    return () => {
      document.removeEventListener('wheel', handleWheel);
    };
  });

  /**
   * Scaling
   * - 触控板 pinch（onWheel ctrl）
   * - 触控板 + meta（onWheel + meta）
   * - 鼠标滚轮 + ctrl（onWheel ctrl）
   */
  React.useEffect(() => {
    const handleWheel = event => {
      event.preventDefault();
      const {
        deltaY,
        ctrlKey,
        metaKey
      } = event;
      if (metaKey || ctrlKey) {
        /**
         * 1. ctrl 或者 cmd 加 wheel 进行缩放
         * 2. 触控板 pinch 缩放，浏览器默认会加上 ctrlKey true
         */
        scaleByOrigin({
          origin: vec2.fromValues(event.clientX, event.clientY),
          getNextScale: currentScale => currentScale * (CURRENT_SCALE_OF_ORIGINAL_SIZE - deltaY * DELTA_Y_SCALE)
        });
      }
    };
    document.addEventListener('wheel', handleWheel, {
      passive: false
    });
    return () => {
      document.removeEventListener('wheel', handleWheel);
    };
  }, [scaleByOrigin]);
  const mobileScalingHandler = useImageViewerMobileScaling();
  const imageProps = {
    onLoad: () => {
      initTransform({
        fitToContainer: true
      });
      setIsImageVisible(true);
    },
    onPointerDown: useEventCallback(composeEventHandlers(singleTouchTranslationHandler, mobileScalingHandler))
  };
  return {
    slotsProps: {
      image: imageProps
    },
    state: {
      isImageVisible,
      setIsImageVisible,
      prevSrc,
      setPrevSrc,
      isPressing
    }
  };
};
export const ImageViewerImage = forwardRef((props, ref) => {
  var _image$alt;
  const {
    errorFallback
  } = props;
  const {
    rootCssScope,
    imageRef,
    image,
    isImageContained
  } = useImageViewerContext();
  const {
    state: {
      status
    },
    slotsProps: useImageSlotsProps
  } = useImage_unstable(image, useMergedRefs(ref, imageRef));
  const {
    slotsProps: useImageViewerImageSlotsProps,
    state: useImageViewerImageState
  } = useImageViewerImage();
  const {
    isImageVisible,
    setIsImageVisible,
    prevSrc,
    setPrevSrc,
    isPressing
  } = useImageViewerImageState;
  const styles = useImageViewerImageStylexStyles({
    prefix: rootCssScope('__image')
  });
  if (image.src !== prevSrc) {
    setIsImageVisible(false);
    setPrevSrc(image.src);
  }

  // 获取光标形状
  const cursor = (() => {
    if (isImageContained) {
      return 'zoom-out';
    }
    if (isPressing) {
      return 'grabbing';
    }
    return 'grab';
  })();
  return jsx(React.Fragment, null, jsx("img", _extends({}, useImageSlotsProps.root, useImageViewerImageSlotsProps.image, {
    sx: [useImageSlotsProps.root.sx, styles.image, status === 'error' && styles.imageError, !isImageVisible && styles.hideImage],
    style: {
      cursor,
      ...useImageSlotsProps.root.style
    },
    alt: (_image$alt = image === null || image === void 0 ? void 0 : image.alt) !== null && _image$alt !== void 0 ? _image$alt : 'preview',
    onLoad: composeEventHandlers(useImageSlotsProps.root.onLoad, useImageViewerImageSlotsProps.image.onLoad)
  })), status === 'error' && resolveSlot(errorFallback, {
    defaultNode: jsx("div", {
      sx: [styles.error]
    }, jsx(LoadfailFilled, {
      className: stylex(styles.errorIcon)
    }))
  }), status === 'loading' && jsx("div", {
    sx: [styles.loading]
  }, jsx(Loading, {
    size: "large"
  })));
});
