// @universe-design/react/es/image-viewer/ImageViewerToolbar.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.urkrog8{width:1px;}
import utkiswu from "./../.stylex/utkiswu";
//.ujamq6b{height:20px;}
import u1ny05fg from "./../.stylex/u1ny05fg";
//.u11fluvs{background:@udtoken-dropdown-button-divider;}
import u256sz from "./../.stylex/u256sz";
//.u1xqy1su{user-select:none;}
import u1443mj0 from "./../.stylex/u1443mj0";
//.u1euv43f{position:absolute;}
import u1u1p7mj from "./../.stylex/u1u1p7mj";
//.u10cq58m{bottom:16px;}
import u1l13cd8 from "./../.stylex/u1l13cd8";
//.u22iagw{display:flex;}
import ukchsq7 from "./../.stylex/ukchsq7";
//.u122n59{align-items:center;}
import ubyhozq from "./../.stylex/ubyhozq";
//.u4d9j23{justify-content:center;}
import ut7q8es from "./../.stylex/ut7q8es";
//.uf2sm71{height:48px;}
import us2xlyu from "./../.stylex/us2xlyu";
//.u1ahpp82{width:max-content;}
import ukjz3g from "./../.stylex/ukjz3g";
//.u198b2fy{background-color:@udtoken-imageview-toolbox-bg;}
import u19bk9ok from "./../.stylex/u19bk9ok";
//.u19ibtj8{color:@static-white;}
import uathu2q from "./../.stylex/uathu2q";
//.ugvjd9r{border-top-left-radius:@border-radius-xs;}
import urnexrx from "./../.stylex/urnexrx";
//.uns275c{border-top-right-radius:@border-radius-xs;}
import u12hcvx2 from "./../.stylex/u12hcvx2";
//.udbqa1p{border-bottom-right-radius:@border-radius-xs;}
import u1gja64x from "./../.stylex/u1gja64x";
//.u1vdvwtp{border-bottom-left-radius:@border-radius-xs;}
import u162mx8s from "./../.stylex/u162mx8s";
//.u1ktbn1t{left:50%;}
import u1jmni06 from "./../.stylex/u1jmni06";
//.u13gfj74{transform:translateX(-50%);}
import u9yw1hi from "./../.stylex/u9yw1hi";
//.u1cguypg{pointer-events:auto;}
import u1ksa1gg from "./../.stylex/u1ksa1gg";
//.u1g0x7ka{padding-top:0;}
import u1fdqv2r from "./../.stylex/u1fdqv2r";
//.u11qrl6u{padding-right:12px;}
import u143d86t from "./../.stylex/u143d86t";
//.u1qch9an{padding-bottom:0;}
import u6qvdto from "./../.stylex/u6qvdto";
//.ujlbh76{padding-left:12px;}
import urjm3ap from "./../.stylex/urjm3ap";
import React from "react";
/* eslint-disable react/no-unknown-property */
/* eslint-disable react/jsx-no-leaked-render */
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx, makeStylexStyles } from '@universe-design/stylex/react';
import stylex from '@universe-design/stylex';
import { useImageViewerContext } from './ImageViewerContext';
import { forwardHTMLRef } from '../util/forwardRef';
import { renderSlotNode, resolveSlot } from '../shared/slot';
import { Space } from '../space';
import { Pagination } from './components/Pagination';
import { Scaling } from './components/Scaling';
import { Rotation } from './components/Rotation';
import { Download } from './components/Download';
import './style';
const useImageViewerToolbarStylexStyles = makeStylexStyles(_ref => {
  let {
    prefix
  } = _ref;
  return {
    toolbar: {
      "props": {
        "famaaq": u1443mj0,
        "qhf8xq": u1u1p7mj,
        "B5kzvoi": u1l13cd8,
        "mc9l5x": ukchsq7,
        "Bt984gj": ubyhozq,
        "Brf1p80": ut7q8es,
        "Bqenvij": us2xlyu,
        "a9b677": ukjz3g,
        "De3pzq": u19bk9ok,
        "sj55zd": uathu2q,
        "Btl43ni": urnexrx,
        "B7oj6ja": u12hcvx2,
        "Bbmb7ep": u1gja64x,
        "Beyfa6y": u162mx8s,
        "oyh7mz": u1jmni06,
        "Bz10aip": u9yw1hi,
        "Bkecrkj": u1ksa1gg,
        "z8tnut": u1fdqv2r,
        "z189sj": u143d86t,
        "Byoj8tv": u6qvdto,
        "uwmqm3": urjm3ap
      },
      "alias": prefix
    },
    toolbarDivider: {
      "props": {
        "a9b677": utkiswu,
        "Bqenvij": u1ny05fg,
        "ayd6f0": u256sz
      },
      "alias": `${prefix}__divider`
    }
  };
});
export const ImageViewerToolbar = forwardHTMLRef((props, ref) => {
  const {
    pagination,
    scaling,
    rotation,
    download: downloadProp,
    children,
    ...htmlProps
  } = props;
  const {
    rootCssScope
  } = useImageViewerContext();
  const styles = useImageViewerToolbarStylexStyles({
    prefix: rootCssScope('__toolbar')
  });
  const downloadNode = renderSlotNode(resolveSlot(downloadProp === true ? {} : downloadProp === false ? null : downloadProp, {
    elementType: Download,
    renderByDefault: true
  }));
  return jsx("div", _extends({
    sx: [styles.toolbar]
  }, htmlProps, {
    ref: ref
  }), jsx(Space, {
    size: 24,
    split: jsx("div", {
      className: stylex(styles.toolbarDivider)
    })
  }, pagination && jsx(Pagination, null), scaling && jsx(Scaling, null), (rotation || downloadNode) && jsx(Space, {
    size: 12
  }, rotation && jsx(Rotation, null), downloadNode), children));
});
