// @universe-design/react/es/image-viewer/ImageViewerDialog.js
//.uly5x3f{width:100%;}
import u13d3rh8 from "./../.stylex/u13d3rh8";
//.u1l02sjl{height:100%;}
import udg2fzc from "./../.stylex/udg2fzc";
//.u1euv43f{position:absolute;}
import u1u1p7mj from "./../.stylex/u1u1p7mj";
/* eslint-disable react/no-unknown-property */
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx } from '@universe-design/stylex/react';
import * as React from 'react';
import stylex from '@universe-design/stylex';
import { Dialog, DialogCancelType, DialogLayoutMode } from '../dialog/dialog';
import { ImageViewerCloseAction } from './useImageViewerVisible';
import { IconButton } from '../button';
import CloseOutlined from '@universe-design/icons-react/CloseOutlined';
import { useImageViewerContext } from './ImageViewerContext';
import './style';
export const ImageViewerDialog = /*#__PURE__*/React.forwardRef((props, ref) => {
  // TODO 原则上来说这块也可以有一个 useImageViewerDialogProps 的 hooks 专门预处理 props，我们的 render 函数体里要尽可能干净
  const {
    zIndex,
    maskClosable = true,
    keyboardClosable,
    hasCloseButton = true,
    children
  } = props;
  const {
    rootCssScope,
    containerRef,
    dialogVisible,
    setDialogVisible
  } = useImageViewerContext();
  return jsx(DialogLayoutMode, {
    mode: "2.18.0"
  }, jsx(Dialog, {
    ref: ref,
    zIndex: zIndex,
    safeMargin: {
      hmax: 0,
      hmin: 0,
      top: 0,
      bottom: 0
    },
    width: "100%",
    wrapperClassName: rootCssScope('__dialog__wrap'),
    destroyOnClose: true,
    keyboardClosable: keyboardClosable,
    overlayClosable: maskClosable,
    overlayClassName: rootCssScope('__dialog__mask'),
    visible: dialogVisible
    // dialog mask 和 esc 关闭
    ,
    onClose: (e, _ref) => {
      let {
        cancelType
      } = _ref;
      if (cancelType === DialogCancelType.Esc) {
        setDialogVisible(false, {
          visible: false,
          event: e,
          action: ImageViewerCloseAction.Esc
        });
        return;
      }
      if (cancelType === DialogCancelType.Mask) {
        setDialogVisible(false, {
          visible: false,
          event: e,
          action: ImageViewerCloseAction.Mask
        });
      }
    }
  }, jsx("div", {
    ref: containerRef,
    sx: [{
      "props": {
        "a9b677": u13d3rh8,
        "Bqenvij": udg2fzc,
        "qhf8xq": u1u1p7mj
      },
      "alias": rootCssScope()
    }]
  }, children, hasCloseButton && jsx(IconButton, {
    size: 20,
    icon: jsx(CloseOutlined, null),
    className: rootCssScope('__close-button'),
    onClick: ev => {
      setDialogVisible(false, {
        visible: false,
        event: ev,
        action: ImageViewerCloseAction.CloseButton
      });
    }
  }))));
});
