// @universe-design/react/es/image-viewer/ImageViewer.js
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx } from '@universe-design/stylex/react';
import * as React from 'react';
import { ImageViewerToolbar } from './ImageViewerToolbar';
import { ImageViewerContextProvider } from './ImageViewerContext';
import { ImageViewerDialog } from './ImageViewerDialog';
import { ImageViewerImage } from './ImageViewerImage';
import { useImageViewer } from './useImageViewer';
import { renderSlotNode, resolveSlot } from '../shared/slot';
import { TOOLBAR_DEFAULT, TOOLBAR_FALSE } from './constants';
import './style';
export const ImageViewer = props => {
  const {
    errorFallback,
    toolbar: toolbarProp = true,
    plugins
  } = props;
  const context = useImageViewer(props);
  const toolbarSlot = resolveSlot((() => {
    if (toolbarProp === true) {
      return TOOLBAR_DEFAULT;
    }
    if (toolbarProp === false) {
      return TOOLBAR_FALSE;
    }
    return toolbarProp;
  })(), {
    elementType: ImageViewerToolbar,
    defaultProps: {
      pagination: TOOLBAR_DEFAULT.pagination,
      scaling: TOOLBAR_DEFAULT.scaling,
      rotation: TOOLBAR_DEFAULT.rotation,
      download: TOOLBAR_DEFAULT.download
    },
    renderByDefault: true
  });
  return jsx(ImageViewerContextProvider, {
    value: context
  }, jsx(ImageViewerDialog, props, jsx(ImageViewerImage, {
    errorFallback
  }), renderSlotNode(toolbarSlot, plugins)));
};
