// @universe-design/react/es/image-viewer/transform.js
import { mat2d, vec2 } from 'gl-matrix';
// eslint-disable-next-line @typescript-eslint/ban-ts-comment, @typescript-eslint/prefer-ts-expect-error
// @ts-ignore mathjs 是 esm 类型和现在的 ts 不兼容
import { lusolve } from 'mathjs';

/**
 * mutable 方法，可以对变换后的 image rect 进行限制
 */
export const clampPoints = (nextPoints, _ref) => {
  let {
    containerRect
  } = _ref;
  const rect = pointsToRect(nextPoints);
  const {
    left,
    right,
    top,
    bottom,
    width,
    height
  } = rect;
  let deltaX = 0;
  let deltaY = 0;
  // 如果图的尺寸小于容器，则位置不动
  if (containerRect.width >= width) {
    deltaX = (containerRect.width - width) / 2 - left;
  } else {
    // 限制图必须贴着边，不能随意拖动
    if (left > containerRect.left) {
      deltaX = containerRect.left - left;
    }
    if (right < containerRect.right) {
      deltaX = containerRect.right - width - left;
    }
  }
  if (containerRect.height >= height) {
    deltaY = (containerRect.height - height) / 2 - top;
  } else {
    if (top > containerRect.top) {
      deltaY = containerRect.top - top;
    }
    if (bottom < containerRect.bottom) {
      deltaY = containerRect.bottom - height - top;
    }
  }
  return nextPoints.map(p => vec2.add(vec2.create(), p, [deltaX, deltaY]));
};

/**
 * 解线性方程得到两个三角之间的变化矩阵
 */
export const getTransformBetweenTwoRects = (points1, points2) => {
  // 我们可以通过解线性方程组来找到这个变换
  const equations = points1.map(point => [[point[0], 0, point[1], 0, 1, 0], [0, point[0], 0, point[1], 0, 1]]).flat(1);
  const solution = lusolve(equations, points2.map(point => [point[0], point[1]]).flat(1));

  // 创建一个 mat2d 矩阵
  const transform = mat2d.fromValues(solution[0][0], solution[1][0], solution[2][0], solution[3][0], solution[4][0], solution[5][0]);
  return transform;
};
export const rectToPoints = rect => [vec2.fromValues(rect.x, rect.y), vec2.fromValues(rect.x + rect.width, rect.y), vec2.fromValues(rect.x, rect.y + rect.height)];
export const pointsToRect = points => {
  const left = Math.min(...points.map(p => p[0]));
  const right = Math.max(...points.map(p => p[0]));
  const top = Math.min(...points.map(p => p[1]));
  const bottom = Math.max(...points.map(p => p[1]));
  const rect = new DOMRect(left, top, right - left, bottom - top);
  return rect;
};
export const getMatScale = mat => Math.hypot(mat[0], mat[1]);
export const getMatRotationRadians = mat => Math.atan2(mat[1], mat[0]);

/**
 * 判断点是否在矩形内
 *
 * @param rect - 矩形区域
 * @param points - 点集合
 * @returns
 */
export const isPointsInRectangleRegion = (rect, points) => {
  // 获取矩形边界
  const rectLeft = rect.left;
  const rectRight = rect.left + rect.width;
  const rectTop = rect.top;
  const rectBottom = rect.top + rect.height;

  // 检查所有点是否在矩形内
  return points.every(point => point[0] >= rectLeft && point[0] <= rectRight && point[1] >= rectTop && point[1] <= rectBottom);
};
