// @universe-design/react/es/image-viewer/useImageViewerVisible.js
import { useControlledState } from '@universe-design/react-utilities';
import { isFunction } from '../util';
import { useEventCallback } from '../hooks';

/**
 * ================================
 *
 * useImageViewerVisible
 *
 * ================================
 */

/**
 * 与图片查看器弹层控制相关的数据逻辑
 */
export const useImageViewerVisible = props => {
  var _visible$defaultValue;
  const {
    visible
  } = props;
  const [dialogVisible, setDialogVisible] = useControlledState((_visible$defaultValue = visible === null || visible === void 0 ? void 0 : visible.defaultValue) !== null && _visible$defaultValue !== void 0 ? _visible$defaultValue : false, visible === null || visible === void 0 ? void 0 : visible.value);
  const setDialogVisibleWithMeta = useEventCallback((value, meta) => {
    setDialogVisible(value);
    if (!visible) {
      return;
    }
    const {
      onChange
    } = visible;
    isFunction(onChange) && onChange(value, meta);
  });
  return {
    state: {
      dialogVisible,
      setDialogVisible: setDialogVisibleWithMeta
    }
  };
};

/**
 * 图片查看器可见性相关属性
 */

/** 关闭图片查看器的参数 */

/**
 * 图片查看器关闭动作
 */
export let ImageViewerCloseAction = /*#__PURE__*/function (ImageViewerCloseAction) {
  ImageViewerCloseAction["Mask"] = "mask";
  ImageViewerCloseAction["CloseButton"] = "closeButton";
  ImageViewerCloseAction["Esc"] = "esc";
  ImageViewerCloseAction["Image"] = "image";
  return ImageViewerCloseAction;
}({});
