// @universe-design/styles/cjs/preset-dark.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// autogen by ../scripts/gen-ts-from-json.ts
const preset_dark_atom_1 = require("./preset-dark.atom");
exports.default = {
    'AI-cursor': preset_dark_atom_1.AI_cursor,
    'AI-dynamic-line': preset_dark_atom_1.AI_dynamic_line,
    'AI-icon-gradient': preset_dark_atom_1.AI_icon_gradient,
    'AI-icon-in-button': preset_dark_atom_1.AI_icon_in_button,
    'AI-loading': preset_dark_atom_1.AI_loading,
    'AI-primary-content-default': preset_dark_atom_1.AI_primary_content_default,
    'AI-primary-content-hover': preset_dark_atom_1.AI_primary_content_hover,
    'AI-primary-content-loading': preset_dark_atom_1.AI_primary_content_loading,
    'AI-primary-content-pressed': preset_dark_atom_1.AI_primary_content_pressed,
    'AI-primary-fill-default': preset_dark_atom_1.AI_primary_fill_default,
    'AI-primary-fill-hover': preset_dark_atom_1.AI_primary_fill_hover,
    'AI-primary-fill-loading': preset_dark_atom_1.AI_primary_fill_loading,
    'AI-primary-fill-pressed': preset_dark_atom_1.AI_primary_fill_pressed,
    'AI-primary-fill-solid-01': preset_dark_atom_1.AI_primary_fill_solid_01,
    'AI-primary-fill-solid-02': preset_dark_atom_1.AI_primary_fill_solid_02,
    'AI-primary-fill-solid-03': preset_dark_atom_1.AI_primary_fill_solid_03,
    'AI-primary-fill-transparent-01': preset_dark_atom_1.AI_primary_fill_transparent_01,
    'AI-primary-fill-transparent-02': preset_dark_atom_1.AI_primary_fill_transparent_02,
    'AI-primary-fill-transparent-03': preset_dark_atom_1.AI_primary_fill_transparent_03,
    'AI-sendicon': preset_dark_atom_1.AI_sendicon,
    'Aurora-bg-main': preset_dark_atom_1.Aurora_bg_main,
    'Aurora-bg-reflection': preset_dark_atom_1.Aurora_bg_reflection,
    'Aurora-bg-subordinate': preset_dark_atom_1.Aurora_bg_subordinate,
    B100: preset_dark_atom_1.B100,
    B200: preset_dark_atom_1.B200,
    B300: preset_dark_atom_1.B300,
    B350: preset_dark_atom_1.B350,
    B400: preset_dark_atom_1.B400,
    B50: preset_dark_atom_1.B50,
    B500: preset_dark_atom_1.B500,
    B600: preset_dark_atom_1.B600,
    B700: preset_dark_atom_1.B700,
    B800: preset_dark_atom_1.B800,
    B900: preset_dark_atom_1.B900,
    C100: preset_dark_atom_1.C100,
    C200: preset_dark_atom_1.C200,
    C300: preset_dark_atom_1.C300,
    C350: preset_dark_atom_1.C350,
    C400: preset_dark_atom_1.C400,
    C50: preset_dark_atom_1.C50,
    C500: preset_dark_atom_1.C500,
    C600: preset_dark_atom_1.C600,
    C700: preset_dark_atom_1.C700,
    C800: preset_dark_atom_1.C800,
    C900: preset_dark_atom_1.C900,
    G100: preset_dark_atom_1.G100,
    G200: preset_dark_atom_1.G200,
    G300: preset_dark_atom_1.G300,
    G350: preset_dark_atom_1.G350,
    G400: preset_dark_atom_1.G400,
    G50: preset_dark_atom_1.G50,
    G500: preset_dark_atom_1.G500,
    G600: preset_dark_atom_1.G600,
    G700: preset_dark_atom_1.G700,
    G800: preset_dark_atom_1.G800,
    G900: preset_dark_atom_1.G900,
    I100: preset_dark_atom_1.I100,
    I200: preset_dark_atom_1.I200,
    I300: preset_dark_atom_1.I300,
    I350: preset_dark_atom_1.I350,
    I400: preset_dark_atom_1.I400,
    I50: preset_dark_atom_1.I50,
    I500: preset_dark_atom_1.I500,
    I600: preset_dark_atom_1.I600,
    I700: preset_dark_atom_1.I700,
    I800: preset_dark_atom_1.I800,
    I900: preset_dark_atom_1.I900,
    L100: preset_dark_atom_1.L100,
    L200: preset_dark_atom_1.L200,
    L300: preset_dark_atom_1.L300,
    L350: preset_dark_atom_1.L350,
    L400: preset_dark_atom_1.L400,
    L50: preset_dark_atom_1.L50,
    L500: preset_dark_atom_1.L500,
    L600: preset_dark_atom_1.L600,
    L700: preset_dark_atom_1.L700,
    L800: preset_dark_atom_1.L800,
    L900: preset_dark_atom_1.L900,
    N00: preset_dark_atom_1.N00,
    'N00-10': preset_dark_atom_1.N00_10,
    'N00-15': preset_dark_atom_1.N00_15,
    'N00-20': preset_dark_atom_1.N00_20,
    'N00-30': preset_dark_atom_1.N00_30,
    'N00-40': preset_dark_atom_1.N00_40,
    'N00-5': preset_dark_atom_1.N00_5,
    'N00-50': preset_dark_atom_1.N00_50,
    'N00-60': preset_dark_atom_1.N00_60,
    'N00-70': preset_dark_atom_1.N00_70,
    'N00-80': preset_dark_atom_1.N00_80,
    'N00-90': preset_dark_atom_1.N00_90,
    N100: preset_dark_atom_1.N100,
    N1000: preset_dark_atom_1.N1000,
    N200: preset_dark_atom_1.N200,
    N300: preset_dark_atom_1.N300,
    N350: preset_dark_atom_1.N350,
    N400: preset_dark_atom_1.N400,
    N50: preset_dark_atom_1.N50,
    N500: preset_dark_atom_1.N500,
    N600: preset_dark_atom_1.N600,
    N650: preset_dark_atom_1.N650,
    N700: preset_dark_atom_1.N700,
    N800: preset_dark_atom_1.N800,
    N900: preset_dark_atom_1.N900,
    'N900-10': preset_dark_atom_1.N900_10,
    'N900-15': preset_dark_atom_1.N900_15,
    'N900-20': preset_dark_atom_1.N900_20,
    'N900-30': preset_dark_atom_1.N900_30,
    'N900-40': preset_dark_atom_1.N900_40,
    'N900-5': preset_dark_atom_1.N900_5,
    'N900-50': preset_dark_atom_1.N900_50,
    'N900-60': preset_dark_atom_1.N900_60,
    'N900-70': preset_dark_atom_1.N900_70,
    'N900-80': preset_dark_atom_1.N900_80,
    'N900-90': preset_dark_atom_1.N900_90,
    N950: preset_dark_atom_1.N950,
    O100: preset_dark_atom_1.O100,
    O200: preset_dark_atom_1.O200,
    O300: preset_dark_atom_1.O300,
    O350: preset_dark_atom_1.O350,
    O400: preset_dark_atom_1.O400,
    O50: preset_dark_atom_1.O50,
    O500: preset_dark_atom_1.O500,
    O600: preset_dark_atom_1.O600,
    O700: preset_dark_atom_1.O700,
    O800: preset_dark_atom_1.O800,
    O900: preset_dark_atom_1.O900,
    P100: preset_dark_atom_1.P100,
    P200: preset_dark_atom_1.P200,
    P300: preset_dark_atom_1.P300,
    P350: preset_dark_atom_1.P350,
    P400: preset_dark_atom_1.P400,
    P50: preset_dark_atom_1.P50,
    P500: preset_dark_atom_1.P500,
    P600: preset_dark_atom_1.P600,
    P700: preset_dark_atom_1.P700,
    P800: preset_dark_atom_1.P800,
    P900: preset_dark_atom_1.P900,
    R100: preset_dark_atom_1.R100,
    R200: preset_dark_atom_1.R200,
    R300: preset_dark_atom_1.R300,
    R350: preset_dark_atom_1.R350,
    R400: preset_dark_atom_1.R400,
    R50: preset_dark_atom_1.R50,
    R500: preset_dark_atom_1.R500,
    R600: preset_dark_atom_1.R600,
    R700: preset_dark_atom_1.R700,
    R800: preset_dark_atom_1.R800,
    R900: preset_dark_atom_1.R900,
    S100: preset_dark_atom_1.S100,
    S200: preset_dark_atom_1.S200,
    S300: preset_dark_atom_1.S300,
    S350: preset_dark_atom_1.S350,
    S400: preset_dark_atom_1.S400,
    S50: preset_dark_atom_1.S50,
    S500: preset_dark_atom_1.S500,
    S600: preset_dark_atom_1.S600,
    S700: preset_dark_atom_1.S700,
    S800: preset_dark_atom_1.S800,
    S900: preset_dark_atom_1.S900,
    T100: preset_dark_atom_1.T100,
    T200: preset_dark_atom_1.T200,
    T300: preset_dark_atom_1.T300,
    T350: preset_dark_atom_1.T350,
    T400: preset_dark_atom_1.T400,
    T50: preset_dark_atom_1.T50,
    T500: preset_dark_atom_1.T500,
    T600: preset_dark_atom_1.T600,
    T700: preset_dark_atom_1.T700,
    T800: preset_dark_atom_1.T800,
    T900: preset_dark_atom_1.T900,
    V100: preset_dark_atom_1.V100,
    V200: preset_dark_atom_1.V200,
    V300: preset_dark_atom_1.V300,
    V350: preset_dark_atom_1.V350,
    V400: preset_dark_atom_1.V400,
    V50: preset_dark_atom_1.V50,
    V500: preset_dark_atom_1.V500,
    V600: preset_dark_atom_1.V600,
    V700: preset_dark_atom_1.V700,
    V800: preset_dark_atom_1.V800,
    V900: preset_dark_atom_1.V900,
    'Vnext-bg-content-header': preset_dark_atom_1.Vnext_bg_content_header,
    'Vnext-bg-gradient': preset_dark_atom_1.Vnext_bg_gradient,
    'Vnext-bg-sub-navigation': preset_dark_atom_1.Vnext_bg_sub_navigation,
    'Vnext-bg-thirdly-navigation': preset_dark_atom_1.Vnext_bg_thirdly_navigation,
    W100: preset_dark_atom_1.W100,
    W200: preset_dark_atom_1.W200,
    W300: preset_dark_atom_1.W300,
    W350: preset_dark_atom_1.W350,
    W400: preset_dark_atom_1.W400,
    W50: preset_dark_atom_1.W50,
    W500: preset_dark_atom_1.W500,
    W600: preset_dark_atom_1.W600,
    W700: preset_dark_atom_1.W700,
    W800: preset_dark_atom_1.W800,
    W900: preset_dark_atom_1.W900,
    Y100: preset_dark_atom_1.Y100,
    Y200: preset_dark_atom_1.Y200,
    Y300: preset_dark_atom_1.Y300,
    Y350: preset_dark_atom_1.Y350,
    Y400: preset_dark_atom_1.Y400,
    Y50: preset_dark_atom_1.Y50,
    Y500: preset_dark_atom_1.Y500,
    Y600: preset_dark_atom_1.Y600,
    Y700: preset_dark_atom_1.Y700,
    Y800: preset_dark_atom_1.Y800,
    Y900: preset_dark_atom_1.Y900,
    'bg-base': preset_dark_atom_1.bg_base,
    'bg-body': preset_dark_atom_1.bg_body,
    'bg-body-overlay': preset_dark_atom_1.bg_body_overlay,
    'bg-content-base': preset_dark_atom_1.bg_content_base,
    'bg-filler': preset_dark_atom_1.bg_filler,
    'bg-float': preset_dark_atom_1.bg_float,
    'bg-float-base': preset_dark_atom_1.bg_float_base,
    'bg-float-overlay': preset_dark_atom_1.bg_float_overlay,
    'bg-float-push': preset_dark_atom_1.bg_float_push,
    'bg-mask': preset_dark_atom_1.bg_mask,
    'bg-pricolor': preset_dark_atom_1.bg_pricolor,
    'bg-sub-navigation': preset_dark_atom_1.bg_sub_navigation,
    'bg-text-selection': preset_dark_atom_1.bg_text_selection,
    'bg-tips': preset_dark_atom_1.bg_tips,
    'colorful-blue': preset_dark_atom_1.colorful_blue,
    'colorful-carmine': preset_dark_atom_1.colorful_carmine,
    'colorful-green': preset_dark_atom_1.colorful_green,
    'colorful-indigo': preset_dark_atom_1.colorful_indigo,
    'colorful-lime': preset_dark_atom_1.colorful_lime,
    'colorful-neutral': preset_dark_atom_1.colorful_neutral,
    'colorful-orange': preset_dark_atom_1.colorful_orange,
    'colorful-purple': preset_dark_atom_1.colorful_purple,
    'colorful-red': preset_dark_atom_1.colorful_red,
    'colorful-sunflower': preset_dark_atom_1.colorful_sunflower,
    'colorful-turquoise': preset_dark_atom_1.colorful_turquoise,
    'colorful-violet': preset_dark_atom_1.colorful_violet,
    'colorful-wathet': preset_dark_atom_1.colorful_wathet,
    'colorful-yellow': preset_dark_atom_1.colorful_yellow,
    'fill-active': preset_dark_atom_1.fill_active,
    'fill-disabled': preset_dark_atom_1.fill_disabled,
    'fill-focus': preset_dark_atom_1.fill_focus,
    'fill-hover': preset_dark_atom_1.fill_hover,
    'fill-img-mask': preset_dark_atom_1.fill_img_mask,
    'fill-loading-mask': preset_dark_atom_1.fill_loading_mask,
    'fill-pressed': preset_dark_atom_1.fill_pressed,
    'fill-selected': preset_dark_atom_1.fill_selected,
    'fill-tag': preset_dark_atom_1.fill_tag,
    'function-danger-100': preset_dark_atom_1.function_danger_100,
    'function-danger-200': preset_dark_atom_1.function_danger_200,
    'function-danger-300': preset_dark_atom_1.function_danger_300,
    'function-danger-350': preset_dark_atom_1.function_danger_350,
    'function-danger-400': preset_dark_atom_1.function_danger_400,
    'function-danger-50': preset_dark_atom_1.function_danger_50,
    'function-danger-500': preset_dark_atom_1.function_danger_500,
    'function-danger-600': preset_dark_atom_1.function_danger_600,
    'function-danger-700': preset_dark_atom_1.function_danger_700,
    'function-danger-800': preset_dark_atom_1.function_danger_800,
    'function-danger-900': preset_dark_atom_1.function_danger_900,
    'function-danger-content-default': preset_dark_atom_1.function_danger_content_default,
    'function-danger-content-hover': preset_dark_atom_1.function_danger_content_hover,
    'function-danger-content-loading': preset_dark_atom_1.function_danger_content_loading,
    'function-danger-content-pressed': preset_dark_atom_1.function_danger_content_pressed,
    'function-danger-fill-default': preset_dark_atom_1.function_danger_fill_default,
    'function-danger-fill-hover': preset_dark_atom_1.function_danger_fill_hover,
    'function-danger-fill-loading': preset_dark_atom_1.function_danger_fill_loading,
    'function-danger-fill-pressed': preset_dark_atom_1.function_danger_fill_pressed,
    'function-danger-fill-solid-01': preset_dark_atom_1.function_danger_fill_solid_01,
    'function-danger-fill-solid-02': preset_dark_atom_1.function_danger_fill_solid_02,
    'function-danger-fill-solid-03': preset_dark_atom_1.function_danger_fill_solid_03,
    'function-danger-fill-transparent-01': preset_dark_atom_1.function_danger_fill_transparent_01,
    'function-danger-fill-transparent-02': preset_dark_atom_1.function_danger_fill_transparent_02,
    'function-danger-fill-transparent-03': preset_dark_atom_1.function_danger_fill_transparent_03,
    'function-danger-on-danger-fill': preset_dark_atom_1.function_danger_on_danger_fill,
    'function-info-100': preset_dark_atom_1.function_info_100,
    'function-info-200': preset_dark_atom_1.function_info_200,
    'function-info-300': preset_dark_atom_1.function_info_300,
    'function-info-350': preset_dark_atom_1.function_info_350,
    'function-info-400': preset_dark_atom_1.function_info_400,
    'function-info-50': preset_dark_atom_1.function_info_50,
    'function-info-500': preset_dark_atom_1.function_info_500,
    'function-info-600': preset_dark_atom_1.function_info_600,
    'function-info-700': preset_dark_atom_1.function_info_700,
    'function-info-800': preset_dark_atom_1.function_info_800,
    'function-info-900': preset_dark_atom_1.function_info_900,
    'function-info-content-default': preset_dark_atom_1.function_info_content_default,
    'function-info-content-hover': preset_dark_atom_1.function_info_content_hover,
    'function-info-content-loading': preset_dark_atom_1.function_info_content_loading,
    'function-info-content-pressed': preset_dark_atom_1.function_info_content_pressed,
    'function-info-fill-default': preset_dark_atom_1.function_info_fill_default,
    'function-info-fill-hover': preset_dark_atom_1.function_info_fill_hover,
    'function-info-fill-loading': preset_dark_atom_1.function_info_fill_loading,
    'function-info-fill-pressed': preset_dark_atom_1.function_info_fill_pressed,
    'function-info-fill-solid-01': preset_dark_atom_1.function_info_fill_solid_01,
    'function-info-fill-solid-02': preset_dark_atom_1.function_info_fill_solid_02,
    'function-info-fill-solid-03': preset_dark_atom_1.function_info_fill_solid_03,
    'function-info-fill-transparent-01': preset_dark_atom_1.function_info_fill_transparent_01,
    'function-info-fill-transparent-02': preset_dark_atom_1.function_info_fill_transparent_02,
    'function-info-fill-transparent-03': preset_dark_atom_1.function_info_fill_transparent_03,
    'function-info-on-info-fill': preset_dark_atom_1.function_info_on_info_fill,
    'function-success-100': preset_dark_atom_1.function_success_100,
    'function-success-200': preset_dark_atom_1.function_success_200,
    'function-success-300': preset_dark_atom_1.function_success_300,
    'function-success-350': preset_dark_atom_1.function_success_350,
    'function-success-400': preset_dark_atom_1.function_success_400,
    'function-success-50': preset_dark_atom_1.function_success_50,
    'function-success-500': preset_dark_atom_1.function_success_500,
    'function-success-600': preset_dark_atom_1.function_success_600,
    'function-success-700': preset_dark_atom_1.function_success_700,
    'function-success-800': preset_dark_atom_1.function_success_800,
    'function-success-900': preset_dark_atom_1.function_success_900,
    'function-success-content-default': preset_dark_atom_1.function_success_content_default,
    'function-success-content-hover': preset_dark_atom_1.function_success_content_hover,
    'function-success-content-loading': preset_dark_atom_1.function_success_content_loading,
    'function-success-content-pressed': preset_dark_atom_1.function_success_content_pressed,
    'function-success-fill-default': preset_dark_atom_1.function_success_fill_default,
    'function-success-fill-hover': preset_dark_atom_1.function_success_fill_hover,
    'function-success-fill-loading': preset_dark_atom_1.function_success_fill_loading,
    'function-success-fill-pressed': preset_dark_atom_1.function_success_fill_pressed,
    'function-success-fill-solid-01': preset_dark_atom_1.function_success_fill_solid_01,
    'function-success-fill-solid-02': preset_dark_atom_1.function_success_fill_solid_02,
    'function-success-fill-solid-03': preset_dark_atom_1.function_success_fill_solid_03,
    'function-success-fill-transparent-01': preset_dark_atom_1.function_success_fill_transparent_01,
    'function-success-fill-transparent-02': preset_dark_atom_1.function_success_fill_transparent_02,
    'function-success-fill-transparent-03': preset_dark_atom_1.function_success_fill_transparent_03,
    'function-success-on-success-fill': preset_dark_atom_1.function_success_on_success_fill,
    'function-warning-100': preset_dark_atom_1.function_warning_100,
    'function-warning-200': preset_dark_atom_1.function_warning_200,
    'function-warning-300': preset_dark_atom_1.function_warning_300,
    'function-warning-350': preset_dark_atom_1.function_warning_350,
    'function-warning-400': preset_dark_atom_1.function_warning_400,
    'function-warning-50': preset_dark_atom_1.function_warning_50,
    'function-warning-500': preset_dark_atom_1.function_warning_500,
    'function-warning-600': preset_dark_atom_1.function_warning_600,
    'function-warning-700': preset_dark_atom_1.function_warning_700,
    'function-warning-800': preset_dark_atom_1.function_warning_800,
    'function-warning-900': preset_dark_atom_1.function_warning_900,
    'function-warning-content-default': preset_dark_atom_1.function_warning_content_default,
    'function-warning-content-hover': preset_dark_atom_1.function_warning_content_hover,
    'function-warning-content-loading': preset_dark_atom_1.function_warning_content_loading,
    'function-warning-content-pressed': preset_dark_atom_1.function_warning_content_pressed,
    'function-warning-fill-default': preset_dark_atom_1.function_warning_fill_default,
    'function-warning-fill-hover': preset_dark_atom_1.function_warning_fill_hover,
    'function-warning-fill-loading': preset_dark_atom_1.function_warning_fill_loading,
    'function-warning-fill-pressed': preset_dark_atom_1.function_warning_fill_pressed,
    'function-warning-fill-solid-01': preset_dark_atom_1.function_warning_fill_solid_01,
    'function-warning-fill-solid-02': preset_dark_atom_1.function_warning_fill_solid_02,
    'function-warning-fill-solid-03': preset_dark_atom_1.function_warning_fill_solid_03,
    'function-warning-fill-transparent-01': preset_dark_atom_1.function_warning_fill_transparent_01,
    'function-warning-fill-transparent-02': preset_dark_atom_1.function_warning_fill_transparent_02,
    'function-warning-fill-transparent-03': preset_dark_atom_1.function_warning_fill_transparent_03,
    'function-warning-on-warning-fill': preset_dark_atom_1.function_warning_on_warning_fill,
    'gradient-blue': preset_dark_atom_1.gradient_blue,
    'gradient-carmine': preset_dark_atom_1.gradient_carmine,
    'gradient-green': preset_dark_atom_1.gradient_green,
    'gradient-indigo': preset_dark_atom_1.gradient_indigo,
    'gradient-lime': preset_dark_atom_1.gradient_lime,
    'gradient-orange': preset_dark_atom_1.gradient_orange,
    'gradient-purple': preset_dark_atom_1.gradient_purple,
    'gradient-red': preset_dark_atom_1.gradient_red,
    'gradient-turquoise': preset_dark_atom_1.gradient_turquoise,
    'gradient-violet': preset_dark_atom_1.gradient_violet,
    'gradient-wathet': preset_dark_atom_1.gradient_wathet,
    'gradient-yellow': preset_dark_atom_1.gradient_yellow,
    'icon-disabled': preset_dark_atom_1.icon_disabled,
    'icon-n1': preset_dark_atom_1.icon_n1,
    'icon-n2': preset_dark_atom_1.icon_n2,
    'icon-n3': preset_dark_atom_1.icon_n3,
    'illustration-blue-a': preset_dark_atom_1.illustration_blue_a,
    'illustration-blue-b': preset_dark_atom_1.illustration_blue_b,
    'illustration-blue-c': preset_dark_atom_1.illustration_blue_c,
    'illustration-blue-d': preset_dark_atom_1.illustration_blue_d,
    'illustration-blue-e': preset_dark_atom_1.illustration_blue_e,
    'illustration-neutral-a': preset_dark_atom_1.illustration_neutral_a,
    'illustration-neutral-b': preset_dark_atom_1.illustration_neutral_b,
    'illustration-neutral-c': preset_dark_atom_1.illustration_neutral_c,
    'illustration-neutral-d': preset_dark_atom_1.illustration_neutral_d,
    'illustration-orange': preset_dark_atom_1.illustration_orange,
    'illustration-red': preset_dark_atom_1.illustration_red,
    'illustration-static-black': preset_dark_atom_1.illustration_static_black,
    'illustration-static-white': preset_dark_atom_1.illustration_static_white,
    'illustration-turquoise': preset_dark_atom_1.illustration_turquoise,
    'illustration-wathet': preset_dark_atom_1.illustration_wathet,
    'illustration-yellow': preset_dark_atom_1.illustration_yellow,
    'inverse-primary-content-default': preset_dark_atom_1.inverse_primary_content_default,
    'inverse-primary-content-hover': preset_dark_atom_1.inverse_primary_content_hover,
    'inverse-primary-content-pressed': preset_dark_atom_1.inverse_primary_content_pressed,
    'line-border-card': preset_dark_atom_1.line_border_card,
    'line-border-component': preset_dark_atom_1.line_border_component,
    'line-divider-default': preset_dark_atom_1.line_divider_default,
    'line-divider-module': preset_dark_atom_1.line_divider_module,
    'primary-content-default': preset_dark_atom_1.primary_content_default,
    'primary-content-hover': preset_dark_atom_1.primary_content_hover,
    'primary-content-loading': preset_dark_atom_1.primary_content_loading,
    'primary-content-pressed': preset_dark_atom_1.primary_content_pressed,
    'primary-fill-default': preset_dark_atom_1.primary_fill_default,
    'primary-fill-hover': preset_dark_atom_1.primary_fill_hover,
    'primary-fill-loading': preset_dark_atom_1.primary_fill_loading,
    'primary-fill-pressed': preset_dark_atom_1.primary_fill_pressed,
    'primary-fill-solid-01': preset_dark_atom_1.primary_fill_solid_01,
    'primary-fill-solid-02': preset_dark_atom_1.primary_fill_solid_02,
    'primary-fill-solid-03': preset_dark_atom_1.primary_fill_solid_03,
    'primary-fill-transparent-01': preset_dark_atom_1.primary_fill_transparent_01,
    'primary-fill-transparent-02': preset_dark_atom_1.primary_fill_transparent_02,
    'primary-fill-transparent-03': preset_dark_atom_1.primary_fill_transparent_03,
    'primary-on-primary-fill': preset_dark_atom_1.primary_on_primary_fill,
    'primary-pri-100': preset_dark_atom_1.primary_pri_100,
    'primary-pri-200': preset_dark_atom_1.primary_pri_200,
    'primary-pri-300': preset_dark_atom_1.primary_pri_300,
    'primary-pri-350': preset_dark_atom_1.primary_pri_350,
    'primary-pri-400': preset_dark_atom_1.primary_pri_400,
    'primary-pri-50': preset_dark_atom_1.primary_pri_50,
    'primary-pri-500': preset_dark_atom_1.primary_pri_500,
    'primary-pri-600': preset_dark_atom_1.primary_pri_600,
    'primary-pri-700': preset_dark_atom_1.primary_pri_700,
    'primary-pri-800': preset_dark_atom_1.primary_pri_800,
    'primary-pri-900': preset_dark_atom_1.primary_pri_900,
    'shadow-default': preset_dark_atom_1.shadow_default,
    'shadow-default-lg': preset_dark_atom_1.shadow_default_lg,
    'shadow-default-md': preset_dark_atom_1.shadow_default_md,
    'shadow-default-sm': preset_dark_atom_1.shadow_default_sm,
    'shadow-lg-down': preset_dark_atom_1.shadow_lg_down,
    'shadow-lg-left': preset_dark_atom_1.shadow_lg_left,
    'shadow-lg-right': preset_dark_atom_1.shadow_lg_right,
    'shadow-lg-up': preset_dark_atom_1.shadow_lg_up,
    'shadow-md-down': preset_dark_atom_1.shadow_md_down,
    'shadow-md-left': preset_dark_atom_1.shadow_md_left,
    'shadow-md-right': preset_dark_atom_1.shadow_md_right,
    'shadow-md-up': preset_dark_atom_1.shadow_md_up,
    'shadow-pri': preset_dark_atom_1.shadow_pri,
    'shadow-pri-lg': preset_dark_atom_1.shadow_pri_lg,
    'shadow-pri-lg-down': preset_dark_atom_1.shadow_pri_lg_down,
    'shadow-pri-md': preset_dark_atom_1.shadow_pri_md,
    'shadow-pri-md-down': preset_dark_atom_1.shadow_pri_md_down,
    'shadow-s1-down': preset_dark_atom_1.shadow_s1_down,
    'shadow-s1-down-pri': preset_dark_atom_1.shadow_s1_down_pri,
    'shadow-s1-left': preset_dark_atom_1.shadow_s1_left,
    'shadow-s1-right': preset_dark_atom_1.shadow_s1_right,
    'shadow-s1-up': preset_dark_atom_1.shadow_s1_up,
    'shadow-s2-down': preset_dark_atom_1.shadow_s2_down,
    'shadow-s2-down-pri': preset_dark_atom_1.shadow_s2_down_pri,
    'shadow-s2-left': preset_dark_atom_1.shadow_s2_left,
    'shadow-s2-right': preset_dark_atom_1.shadow_s2_right,
    'shadow-s2-up': preset_dark_atom_1.shadow_s2_up,
    'shadow-s3-down': preset_dark_atom_1.shadow_s3_down,
    'shadow-s3-down-pri': preset_dark_atom_1.shadow_s3_down_pri,
    'shadow-s3-left': preset_dark_atom_1.shadow_s3_left,
    'shadow-s3-right': preset_dark_atom_1.shadow_s3_right,
    'shadow-s3-up': preset_dark_atom_1.shadow_s3_up,
    'shadow-s4-down': preset_dark_atom_1.shadow_s4_down,
    'shadow-s4-down-pri': preset_dark_atom_1.shadow_s4_down_pri,
    'shadow-s4-left': preset_dark_atom_1.shadow_s4_left,
    'shadow-s4-right': preset_dark_atom_1.shadow_s4_right,
    'shadow-s4-up': preset_dark_atom_1.shadow_s4_up,
    'shadow-s5-down': preset_dark_atom_1.shadow_s5_down,
    'shadow-s5-down-pri': preset_dark_atom_1.shadow_s5_down_pri,
    'shadow-s5-left': preset_dark_atom_1.shadow_s5_left,
    'shadow-s5-right': preset_dark_atom_1.shadow_s5_right,
    'shadow-s5-up': preset_dark_atom_1.shadow_s5_up,
    'shadow-sm-down': preset_dark_atom_1.shadow_sm_down,
    'shadow-sm-left': preset_dark_atom_1.shadow_sm_left,
    'shadow-sm-right': preset_dark_atom_1.shadow_sm_right,
    'shadow-sm-up': preset_dark_atom_1.shadow_sm_up,
    'static-black': preset_dark_atom_1.static_black,
    'static-black-10': preset_dark_atom_1.static_black_10,
    'static-black-15': preset_dark_atom_1.static_black_15,
    'static-black-20': preset_dark_atom_1.static_black_20,
    'static-black-30': preset_dark_atom_1.static_black_30,
    'static-black-40': preset_dark_atom_1.static_black_40,
    'static-black-5': preset_dark_atom_1.static_black_5,
    'static-black-50': preset_dark_atom_1.static_black_50,
    'static-black-60': preset_dark_atom_1.static_black_60,
    'static-black-70': preset_dark_atom_1.static_black_70,
    'static-black-80': preset_dark_atom_1.static_black_80,
    'static-black-90': preset_dark_atom_1.static_black_90,
    'static-white': preset_dark_atom_1.static_white,
    'static-white-10': preset_dark_atom_1.static_white_10,
    'static-white-15': preset_dark_atom_1.static_white_15,
    'static-white-20': preset_dark_atom_1.static_white_20,
    'static-white-30': preset_dark_atom_1.static_white_30,
    'static-white-40': preset_dark_atom_1.static_white_40,
    'static-white-5': preset_dark_atom_1.static_white_5,
    'static-white-50': preset_dark_atom_1.static_white_50,
    'static-white-60': preset_dark_atom_1.static_white_60,
    'static-white-70': preset_dark_atom_1.static_white_70,
    'static-white-80': preset_dark_atom_1.static_white_80,
    'static-white-90': preset_dark_atom_1.static_white_90,
    'static-white-hover': preset_dark_atom_1.static_white_hover,
    'static-white-pressed': preset_dark_atom_1.static_white_pressed,
    'text-caption': preset_dark_atom_1.text_caption,
    'text-disabled': preset_dark_atom_1.text_disabled,
    'text-link-disabled': preset_dark_atom_1.text_link_disabled,
    'text-link-hover': preset_dark_atom_1.text_link_hover,
    'text-link-loading': preset_dark_atom_1.text_link_loading,
    'text-link-normal': preset_dark_atom_1.text_link_normal,
    'text-link-pressed': preset_dark_atom_1.text_link_pressed,
    'text-placeholder': preset_dark_atom_1.text_placeholder,
    'text-title': preset_dark_atom_1.text_title,
};
