// @universe-design/styles/cjs/preset-light.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.G400 = exports.G350 = exports.G300 = exports.G200 = exports.G100 = exports.C900 = exports.C800 = exports.C700 = exports.C600 = exports.C500 = exports.C50 = exports.C400 = exports.C350 = exports.C300 = exports.C200 = exports.C100 = exports.B900 = exports.B800 = exports.B700 = exports.B600 = exports.B500 = exports.B50 = exports.B400 = exports.B350 = exports.B300 = exports.B200 = exports.B100 = exports.Aurora_bg_subordinate = exports.Aurora_bg_reflection = exports.Aurora_bg_main = exports.AI_sendicon = exports.AI_primary_fill_transparent_03 = exports.AI_primary_fill_transparent_02 = exports.AI_primary_fill_transparent_01 = exports.AI_primary_fill_solid_03 = exports.AI_primary_fill_solid_02 = exports.AI_primary_fill_solid_01 = exports.AI_primary_fill_pressed = exports.AI_primary_fill_loading = exports.AI_primary_fill_hover = exports.AI_primary_fill_default = exports.AI_primary_content_pressed = exports.AI_primary_content_loading = exports.AI_primary_content_hover = exports.AI_primary_content_default = exports.AI_loading = exports.AI_icon_in_button = exports.AI_icon_gradient = exports.AI_dynamic_line = exports.AI_cursor = void 0;
exports.N650 = exports.N600 = exports.N500 = exports.N50 = exports.N400 = exports.N350 = exports.N300 = exports.N200 = exports.N1000 = exports.N100 = exports.N00_90 = exports.N00_80 = exports.N00_70 = exports.N00_60 = exports.N00_50 = exports.N00_5 = exports.N00_40 = exports.N00_30 = exports.N00_20 = exports.N00_15 = exports.N00_10 = exports.N00 = exports.L900 = exports.L800 = exports.L700 = exports.L600 = exports.L500 = exports.L50 = exports.L400 = exports.L350 = exports.L300 = exports.L200 = exports.L100 = exports.I900 = exports.I800 = exports.I700 = exports.I600 = exports.I500 = exports.I50 = exports.I400 = exports.I350 = exports.I300 = exports.I200 = exports.I100 = exports.G900 = exports.G800 = exports.G700 = exports.G600 = exports.G500 = exports.G50 = void 0;
exports.S200 = exports.S100 = exports.R900 = exports.R800 = exports.R700 = exports.R600 = exports.R500 = exports.R50 = exports.R400 = exports.R350 = exports.R300 = exports.R200 = exports.R100 = exports.P900 = exports.P800 = exports.P700 = exports.P600 = exports.P500 = exports.P50 = exports.P400 = exports.P350 = exports.P300 = exports.P200 = exports.P100 = exports.O900 = exports.O800 = exports.O700 = exports.O600 = exports.O500 = exports.O50 = exports.O400 = exports.O350 = exports.O300 = exports.O200 = exports.O100 = exports.N950 = exports.N900_90 = exports.N900_80 = exports.N900_70 = exports.N900_60 = exports.N900_50 = exports.N900_5 = exports.N900_40 = exports.N900_30 = exports.N900_20 = exports.N900_15 = exports.N900_10 = exports.N900 = exports.N800 = exports.N700 = void 0;
exports.Y350 = exports.Y300 = exports.Y200 = exports.Y100 = exports.W900 = exports.W800 = exports.W700 = exports.W600 = exports.W500 = exports.W50 = exports.W400 = exports.W350 = exports.W300 = exports.W200 = exports.W100 = exports.Vnext_bg_thirdly_navigation = exports.Vnext_bg_sub_navigation = exports.Vnext_bg_gradient = exports.Vnext_bg_content_header = exports.V900 = exports.V800 = exports.V700 = exports.V600 = exports.V500 = exports.V50 = exports.V400 = exports.V350 = exports.V300 = exports.V200 = exports.V100 = exports.T900 = exports.T800 = exports.T700 = exports.T600 = exports.T500 = exports.T50 = exports.T400 = exports.T350 = exports.T300 = exports.T200 = exports.T100 = exports.S900 = exports.S800 = exports.S700 = exports.S600 = exports.S500 = exports.S50 = exports.S400 = exports.S350 = exports.S300 = void 0;
exports.function_danger_50 = exports.function_danger_400 = exports.function_danger_350 = exports.function_danger_300 = exports.function_danger_200 = exports.function_danger_100 = exports.fill_tag = exports.fill_selected = exports.fill_pressed = exports.fill_loading_mask = exports.fill_img_mask = exports.fill_hover = exports.fill_focus = exports.fill_disabled = exports.fill_active = exports.colorful_yellow = exports.colorful_wathet = exports.colorful_violet = exports.colorful_turquoise = exports.colorful_sunflower = exports.colorful_red = exports.colorful_purple = exports.colorful_orange = exports.colorful_neutral = exports.colorful_lime = exports.colorful_indigo = exports.colorful_green = exports.colorful_carmine = exports.colorful_blue = exports.bg_tips = exports.bg_text_selection = exports.bg_sub_navigation = exports.bg_pricolor = exports.bg_mask = exports.bg_float_push = exports.bg_float_overlay = exports.bg_float_base = exports.bg_float = exports.bg_filler = exports.bg_content_base = exports.bg_body_overlay = exports.bg_body = exports.bg_base = exports.Y900 = exports.Y800 = exports.Y700 = exports.Y600 = exports.Y500 = exports.Y50 = exports.Y400 = void 0;
exports.function_success_350 = exports.function_success_300 = exports.function_success_200 = exports.function_success_100 = exports.function_info_on_info_fill = exports.function_info_fill_transparent_03 = exports.function_info_fill_transparent_02 = exports.function_info_fill_transparent_01 = exports.function_info_fill_solid_03 = exports.function_info_fill_solid_02 = exports.function_info_fill_solid_01 = exports.function_info_fill_pressed = exports.function_info_fill_loading = exports.function_info_fill_hover = exports.function_info_fill_default = exports.function_info_content_pressed = exports.function_info_content_loading = exports.function_info_content_hover = exports.function_info_content_default = exports.function_info_900 = exports.function_info_800 = exports.function_info_700 = exports.function_info_600 = exports.function_info_500 = exports.function_info_50 = exports.function_info_400 = exports.function_info_350 = exports.function_info_300 = exports.function_info_200 = exports.function_info_100 = exports.function_danger_on_danger_fill = exports.function_danger_fill_transparent_03 = exports.function_danger_fill_transparent_02 = exports.function_danger_fill_transparent_01 = exports.function_danger_fill_solid_03 = exports.function_danger_fill_solid_02 = exports.function_danger_fill_solid_01 = exports.function_danger_fill_pressed = exports.function_danger_fill_loading = exports.function_danger_fill_hover = exports.function_danger_fill_default = exports.function_danger_content_pressed = exports.function_danger_content_loading = exports.function_danger_content_hover = exports.function_danger_content_default = exports.function_danger_900 = exports.function_danger_800 = exports.function_danger_700 = exports.function_danger_600 = exports.function_danger_500 = void 0;
exports.gradient_carmine = exports.gradient_blue = exports.function_warning_on_warning_fill = exports.function_warning_fill_transparent_03 = exports.function_warning_fill_transparent_02 = exports.function_warning_fill_transparent_01 = exports.function_warning_fill_solid_03 = exports.function_warning_fill_solid_02 = exports.function_warning_fill_solid_01 = exports.function_warning_fill_pressed = exports.function_warning_fill_loading = exports.function_warning_fill_hover = exports.function_warning_fill_default = exports.function_warning_content_pressed = exports.function_warning_content_loading = exports.function_warning_content_hover = exports.function_warning_content_default = exports.function_warning_900 = exports.function_warning_800 = exports.function_warning_700 = exports.function_warning_600 = exports.function_warning_500 = exports.function_warning_50 = exports.function_warning_400 = exports.function_warning_350 = exports.function_warning_300 = exports.function_warning_200 = exports.function_warning_100 = exports.function_success_on_success_fill = exports.function_success_fill_transparent_03 = exports.function_success_fill_transparent_02 = exports.function_success_fill_transparent_01 = exports.function_success_fill_solid_03 = exports.function_success_fill_solid_02 = exports.function_success_fill_solid_01 = exports.function_success_fill_pressed = exports.function_success_fill_loading = exports.function_success_fill_hover = exports.function_success_fill_default = exports.function_success_content_pressed = exports.function_success_content_loading = exports.function_success_content_hover = exports.function_success_content_default = exports.function_success_900 = exports.function_success_800 = exports.function_success_700 = exports.function_success_600 = exports.function_success_500 = exports.function_success_50 = exports.function_success_400 = void 0;
exports.primary_fill_transparent_02 = exports.primary_fill_transparent_01 = exports.primary_fill_solid_03 = exports.primary_fill_solid_02 = exports.primary_fill_solid_01 = exports.primary_fill_pressed = exports.primary_fill_loading = exports.primary_fill_hover = exports.primary_fill_default = exports.primary_content_pressed = exports.primary_content_loading = exports.primary_content_hover = exports.primary_content_default = exports.line_divider_module = exports.line_divider_default = exports.line_border_component = exports.line_border_card = exports.inverse_primary_content_pressed = exports.inverse_primary_content_hover = exports.inverse_primary_content_default = exports.illustration_yellow = exports.illustration_wathet = exports.illustration_turquoise = exports.illustration_static_white = exports.illustration_static_black = exports.illustration_red = exports.illustration_orange = exports.illustration_neutral_d = exports.illustration_neutral_c = exports.illustration_neutral_b = exports.illustration_neutral_a = exports.illustration_blue_e = exports.illustration_blue_d = exports.illustration_blue_c = exports.illustration_blue_b = exports.illustration_blue_a = exports.icon_n3 = exports.icon_n2 = exports.icon_n1 = exports.icon_disabled = exports.gradient_yellow = exports.gradient_wathet = exports.gradient_violet = exports.gradient_turquoise = exports.gradient_red = exports.gradient_purple = exports.gradient_orange = exports.gradient_lime = exports.gradient_indigo = exports.gradient_green = void 0;
exports.shadow_s4_up = exports.shadow_s4_right = exports.shadow_s4_left = exports.shadow_s4_down_pri = exports.shadow_s4_down = exports.shadow_s3_up = exports.shadow_s3_right = exports.shadow_s3_left = exports.shadow_s3_down_pri = exports.shadow_s3_down = exports.shadow_s2_up = exports.shadow_s2_right = exports.shadow_s2_left = exports.shadow_s2_down_pri = exports.shadow_s2_down = exports.shadow_s1_up = exports.shadow_s1_right = exports.shadow_s1_left = exports.shadow_s1_down_pri = exports.shadow_s1_down = exports.shadow_pri_md_down = exports.shadow_pri_md = exports.shadow_pri_lg_down = exports.shadow_pri_lg = exports.shadow_pri = exports.shadow_md_up = exports.shadow_md_right = exports.shadow_md_left = exports.shadow_md_down = exports.shadow_lg_up = exports.shadow_lg_right = exports.shadow_lg_left = exports.shadow_lg_down = exports.shadow_default_sm = exports.shadow_default_md = exports.shadow_default_lg = exports.shadow_default = exports.primary_pri_900 = exports.primary_pri_800 = exports.primary_pri_700 = exports.primary_pri_600 = exports.primary_pri_500 = exports.primary_pri_50 = exports.primary_pri_400 = exports.primary_pri_350 = exports.primary_pri_300 = exports.primary_pri_200 = exports.primary_pri_100 = exports.primary_on_primary_fill = exports.primary_fill_transparent_03 = void 0;
exports.text_title = exports.text_placeholder = exports.text_link_pressed = exports.text_link_normal = exports.text_link_loading = exports.text_link_hover = exports.text_link_disabled = exports.text_disabled = exports.text_caption = exports.static_white_pressed = exports.static_white_hover = exports.static_white_90 = exports.static_white_80 = exports.static_white_70 = exports.static_white_60 = exports.static_white_50 = exports.static_white_5 = exports.static_white_40 = exports.static_white_30 = exports.static_white_20 = exports.static_white_15 = exports.static_white_10 = exports.static_white = exports.static_black_90 = exports.static_black_80 = exports.static_black_70 = exports.static_black_60 = exports.static_black_50 = exports.static_black_5 = exports.static_black_40 = exports.static_black_30 = exports.static_black_20 = exports.static_black_15 = exports.static_black_10 = exports.static_black = exports.shadow_sm_up = exports.shadow_sm_right = exports.shadow_sm_left = exports.shadow_sm_down = exports.shadow_s5_up = exports.shadow_s5_right = exports.shadow_s5_left = exports.shadow_s5_down_pri = exports.shadow_s5_down = void 0;
// autogen by ../scripts/gen-ts-from-json.ts
exports.AI_cursor = 'rgb(71, 82, 230)';
exports.AI_dynamic_line = 'conic-gradient(from 45deg, rgb(207, 94, 207) 0%, rgb(71, 82, 230) 33.33333333333333%, rgb(71, 82, 230) 66.66666666666666%, rgb(207, 94, 207) 100%)';
exports.AI_icon_gradient = 'linear-gradient(to right, rgb(71, 82, 230) 0%, rgb(159, 111, 241) 100%)';
exports.AI_icon_in_button = 'rgb(71, 82, 230)';
exports.AI_loading = 'linear-gradient(to right, rgb(71, 82, 230) 0%, rgb(222, 129, 222) 100%)';
exports.AI_primary_content_default = 'linear-gradient(to right, rgb(71, 82, 230) 0%, rgb(207, 94, 207) 100%)';
exports.AI_primary_content_hover = 'linear-gradient(to right, rgb(91, 101, 245) 0%, rgb(222, 129, 222) 100%)';
exports.AI_primary_content_loading = 'linear-gradient(to right, rgba(71, 82, 230, 0.6) 0%, rgba(207, 94, 207, 0.6) 100%)';
exports.AI_primary_content_pressed = 'linear-gradient(to right, rgb(71, 82, 230) 0%, rgb(191, 61, 191) 100%)';
exports.AI_primary_fill_default = 'linear-gradient(to right, rgb(91, 101, 245) 0%, rgb(222, 129, 222) 100%)';
exports.AI_primary_fill_hover = 'linear-gradient(to right, rgb(117, 125, 240) 0%, rgb(229, 156, 229) 100%)';
exports.AI_primary_fill_loading = 'linear-gradient(to right, rgba(91, 101, 245, 0.6) 0%, rgba(222, 129, 222, 0.6) 100%)';
exports.AI_primary_fill_pressed = 'linear-gradient(to right, rgb(71, 82, 230) 0%, rgb(207, 94, 207) 100%)';
exports.AI_primary_fill_solid_01 = 'linear-gradient(to right, rgb(242, 243, 253) 0%, rgb(252, 238, 252) 100%)';
exports.AI_primary_fill_solid_02 = 'linear-gradient(to right, rgb(233, 234, 251) 0%, rgb(249, 226, 249) 100%)';
exports.AI_primary_fill_solid_03 = 'linear-gradient(to right, rgb(204, 207, 249) 0%, rgb(243, 196, 243) 100%)';
exports.AI_primary_fill_transparent_01 = 'linear-gradient(to right, rgba(91, 101, 245, 0.08) 0%, rgba(222, 129, 222, 0.08) 100%)';
exports.AI_primary_fill_transparent_02 = 'linear-gradient(to right, rgba(91, 101, 245, 0.2) 0%, rgba(222, 129, 222, 0.2) 100%)';
exports.AI_primary_fill_transparent_03 = 'linear-gradient(to right, rgba(91, 101, 245, 0.3) 0%, rgba(222, 129, 222, 0.3) 100%)';
exports.AI_sendicon = 'linear-gradient(to right, rgb(71, 82, 230) 0%, rgb(159, 111, 241) 100%)';
exports.Aurora_bg_main = 'rgba(20, 86, 240, 0.2)';
exports.Aurora_bg_reflection = 'rgba(45, 190, 171, 0.1)';
exports.Aurora_bg_subordinate = 'rgba(51, 109, 244, 0.2)';
exports.B100 = '#e0e9ff';
exports.B200 = '#c2d4ff';
exports.B300 = '#94b4ff';
exports.B350 = '#7aa2ff';
exports.B400 = '#5083fb';
exports.B50 = '#f0f4ff';
exports.B500 = '#336df4';
exports.B600 = '#1456f0';
exports.B700 = '#0442d2';
exports.B800 = '#002f9e';
exports.B900 = '#002270';
exports.C100 = '#fee2f2';
exports.C200 = '#f8c4e1';
exports.C300 = '#f598cc';
exports.C350 = '#eb78b8';
exports.C400 = '#df58a5';
exports.C50 = '#fef0f8';
exports.C500 = '#cc398c';
exports.C600 = '#b82879';
exports.C700 = '#9d1562';
exports.C800 = '#730744';
exports.C900 = '#550c35';
exports.G100 = '#d0f5ce';
exports.G200 = '#95e599';
exports.G300 = '#5cd168';
exports.G350 = '#35bd4b';
exports.G400 = '#32a645';
exports.G50 = '#e4fae1';
exports.G500 = '#258832';
exports.G600 = '#1a7526';
exports.G700 = '#0b6017';
exports.G800 = '#04430c';
exports.G900 = '#022c07';
exports.I100 = '#e9eafb';
exports.I200 = '#cccff9';
exports.I300 = '#abb0f2';
exports.I350 = '#959bee';
exports.I400 = '#757df0';
exports.I50 = '#f2f3fd';
exports.I500 = '#5b65f5';
exports.I600 = '#4752e6';
exports.I700 = '#333dcc';
exports.I800 = '#1e27a4';
exports.I900 = '#151b70';
exports.L100 = '#e3f0a3';
exports.L200 = '#c8dd5f';
exports.L300 = '#a2c10b';
exports.L350 = '#91ad00';
exports.L400 = '#7b9207';
exports.L50 = '#f2f8d3';
exports.L500 = '#6b7f06';
exports.L600 = '#5c6d08';
exports.L700 = '#4a5804';
exports.L800 = '#333d00';
exports.L900 = '#262e00';
exports.N00 = '#ffffff';
exports.N00_10 = 'rgba(255, 255, 255, 0.1)';
exports.N00_15 = 'rgba(255, 255, 255, 0.15)';
exports.N00_20 = 'rgba(255, 255, 255, 0.2)';
exports.N00_30 = 'rgba(255, 255, 255, 0.3)';
exports.N00_40 = 'rgba(255, 255, 255, 0.4)';
exports.N00_5 = 'rgba(255, 255, 255, 0.05)';
exports.N00_50 = 'rgba(255, 255, 255, 0.5)';
exports.N00_60 = 'rgba(255, 255, 255, 0.6)';
exports.N00_70 = 'rgba(255, 255, 255, 0.7)';
exports.N00_80 = 'rgba(255, 255, 255, 0.8)';
exports.N00_90 = 'rgba(255, 255, 255, 0.9)';
exports.N100 = '#f2f3f5';
exports.N1000 = '#000000';
exports.N200 = '#eff0f1';
exports.N300 = '#dee0e3';
exports.N350 = '#d0d3d6';
exports.N400 = '#bbbfc4';
exports.N50 = '#f5f6f7';
exports.N500 = '#8f959e';
exports.N600 = '#646a73';
exports.N650 = '#51565d';
exports.N700 = '#373c43';
exports.N800 = '#2b2f36';
exports.N900 = '#1f2329';
exports.N900_10 = 'rgba(31, 35, 41, 0.1)';
exports.N900_15 = 'rgba(31, 35, 41, 0.15)';
exports.N900_20 = 'rgba(31, 35, 41, 0.2)';
exports.N900_30 = 'rgba(31, 35, 41, 0.3)';
exports.N900_40 = 'rgba(31, 35, 41, 0.4)';
exports.N900_5 = 'rgba(31, 35, 41, 0.05)';
exports.N900_50 = 'rgba(31, 35, 41, 0.5)';
exports.N900_60 = 'rgba(31, 35, 41, 0.6)';
exports.N900_70 = 'rgba(31, 35, 41, 0.7)';
exports.N900_80 = 'rgba(31, 35, 41, 0.8)';
exports.N900_90 = 'rgba(31, 35, 41, 0.9)';
exports.N950 = '#0f1114';
exports.O100 = '#fee7cd';
exports.O200 = '#fec48b';
exports.O300 = '#ff9d4c';
exports.O350 = '#ff811a';
exports.O400 = '#ed6d0c';
exports.O50 = '#fff3e5';
exports.O500 = '#c25705';
exports.O600 = '#a44904';
exports.O700 = '#853a05';
exports.O800 = '#642b02';
exports.O900 = '#3b1a02';
exports.P100 = '#efe6fe';
exports.P200 = '#dcc9fd';
exports.P300 = '#c8a9fc';
exports.P350 = '#b791fa';
exports.P400 = '#9f6ff1';
exports.P50 = '#f5f0fe';
exports.P500 = '#8d55ed';
exports.P600 = '#7a35f0';
exports.P700 = '#611fd6';
exports.P800 = '#4811a6';
exports.P900 = '#2f0080';
exports.R100 = '#fee3e2';
exports.R200 = '#fdc6c4';
exports.R300 = '#f89e9b';
exports.R350 = '#ff7570';
exports.R400 = '#f54a45';
exports.R50 = '#fef0f0';
exports.R500 = '#e22e28';
exports.R600 = '#c02a26';
exports.R700 = '#a11c17';
exports.R800 = '#741915';
exports.R900 = '#590603';
exports.S100 = '#fffca3';
exports.S200 = '#fff67a';
exports.S300 = '#fff258';
exports.S350 = '#ffe928';
exports.S400 = '#e5ce00';
exports.S50 = '#ffffdb';
exports.S500 = '#c2a800';
exports.S600 = '#8f7c00';
exports.S700 = '#5c4f00';
exports.S800 = '#423700';
exports.S900 = '#2c2502';
exports.T100 = '#c4f2ec';
exports.T200 = '#6fe8d8';
exports.T300 = '#33d6c0';
exports.T350 = '#2dbeab';
exports.T400 = '#10a893';
exports.T50 = '#e2f8f5';
exports.T500 = '#0f8575';
exports.T600 = '#067062';
exports.T700 = '#045d51';
exports.T800 = '#03443b';
exports.T900 = '#02312a';
exports.V100 = '#f9e2f9';
exports.V200 = '#f3c4f3';
exports.V300 = '#e59ce5';
exports.V350 = '#de81de';
exports.V400 = '#cf5ecf';
exports.V50 = '#fceefc';
exports.V500 = '#bf3dbf';
exports.V600 = '#a630a6';
exports.V700 = '#872787';
exports.V800 = '#6a116a';
exports.V900 = '#520052';
exports.Vnext_bg_content_header = 'rgb(252, 252, 252)';
exports.Vnext_bg_gradient = 'linear-gradient(180deg, rgba(255, 255, 255, 0.05) 0%, rgba(100, 106, 115, 0.05) 100%)';
exports.Vnext_bg_sub_navigation = 'rgba(245, 245, 245, 0.7)';
exports.Vnext_bg_thirdly_navigation = 'rgba(252, 252, 252, 0.9)';
exports.W100 = '#caeffc';
exports.W200 = '#97dcfc';
exports.W300 = '#3ec3f7';
exports.W350 = '#25b0e7';
exports.W400 = '#1295ca';
exports.W50 = '#e7f8fe';
exports.W500 = '#047fb0';
exports.W600 = '#076a94';
exports.W700 = '#0f587a';
exports.W800 = '#06415c';
exports.W900 = '#072b3d';
exports.Y100 = '#faedc2';
exports.Y200 = '#fcdf7e';
exports.Y300 = '#fad355';
exports.Y350 = '#ffc60a';
exports.Y400 = '#d99904';
exports.Y50 = '#fbf4df';
exports.Y500 = '#ad7a03';
exports.Y600 = '#865b03';
exports.Y700 = '#6f4a01';
exports.Y800 = '#573601';
exports.Y900 = '#382201';
exports.bg_base = 'rgb(242, 243, 245)';
exports.bg_body = 'rgb(255, 255, 255)';
exports.bg_body_overlay = 'rgb(245, 246, 247)';
exports.bg_content_base = 'rgb(248, 249, 250)';
exports.bg_filler = 'rgb(239, 240, 241)';
exports.bg_float = 'rgb(255, 255, 255)';
exports.bg_float_base = 'rgb(242, 243, 245)';
exports.bg_float_overlay = 'rgb(245, 246, 247)';
exports.bg_float_push = 'rgba(255, 255, 255, 0.8)';
exports.bg_mask = 'rgba(0, 0, 0, 0.55)';
exports.bg_pricolor = 'rgb(80, 131, 251)';
exports.bg_sub_navigation = 'rgb(245, 246, 247)';
exports.bg_text_selection = 'rgba(20, 86, 240, 0.3)';
exports.bg_tips = 'rgb(31, 35, 41)';
exports.colorful_blue = 'rgb(51, 109, 244)';
exports.colorful_carmine = 'rgb(223, 88, 165)';
exports.colorful_green = 'rgb(53, 189, 75)';
exports.colorful_indigo = 'rgb(91, 101, 245)';
exports.colorful_lime = 'rgb(145, 173, 0)';
exports.colorful_neutral = 'rgb(143, 149, 158)';
exports.colorful_orange = 'rgb(255, 129, 26)';
exports.colorful_purple = 'rgb(141, 85, 237)';
exports.colorful_red = 'rgb(245, 74, 69)';
exports.colorful_sunflower = 'rgb(255, 233, 40)';
exports.colorful_turquoise = 'rgb(45, 190, 171)';
exports.colorful_violet = 'rgb(191, 61, 191)';
exports.colorful_wathet = 'rgb(37, 176, 231)';
exports.colorful_yellow = 'rgb(255, 198, 10)';
exports.fill_active = 'rgba(20, 86, 240, 0.15)';
exports.fill_disabled = 'rgb(187, 191, 196)';
exports.fill_focus = 'rgba(31, 35, 41, 0.12)';
exports.fill_hover = 'rgba(31, 35, 41, 0.08)';
exports.fill_img_mask = 'rgba(255, 255, 255, 0)';
exports.fill_loading_mask = 'rgba(255, 255, 255, 0.6)';
exports.fill_pressed = 'rgba(31, 35, 41, 0.12)';
exports.fill_selected = 'rgba(20, 86, 240, 0.1)';
exports.fill_tag = 'rgba(31, 35, 41, 0.1)';
exports.function_danger_100 = 'rgb(254, 227, 226)';
exports.function_danger_200 = 'rgb(253, 198, 196)';
exports.function_danger_300 = 'rgb(248, 158, 155)';
exports.function_danger_350 = 'rgb(255, 117, 112)';
exports.function_danger_400 = 'rgb(245, 74, 69)';
exports.function_danger_50 = 'rgb(254, 240, 240)';
exports.function_danger_500 = 'rgb(226, 46, 40)';
exports.function_danger_600 = 'rgb(192, 42, 38)';
exports.function_danger_700 = 'rgb(161, 28, 23)';
exports.function_danger_800 = 'rgb(116, 25, 21)';
exports.function_danger_900 = 'rgb(89, 6, 3)';
exports.function_danger_content_default = 'rgb(245, 74, 69)';
exports.function_danger_content_hover = 'rgb(255, 117, 112)';
exports.function_danger_content_loading = 'rgb(248, 158, 155)';
exports.function_danger_content_pressed = 'rgb(226, 46, 40)';
exports.function_danger_fill_default = 'rgb(245, 74, 69)';
exports.function_danger_fill_hover = 'rgb(255, 117, 112)';
exports.function_danger_fill_loading = 'rgb(248, 158, 155)';
exports.function_danger_fill_pressed = 'rgb(226, 46, 40)';
exports.function_danger_fill_solid_01 = 'rgb(254, 240, 240)';
exports.function_danger_fill_solid_02 = 'rgb(254, 227, 226)';
exports.function_danger_fill_solid_03 = 'rgb(253, 198, 196)';
exports.function_danger_fill_transparent_01 = 'rgba(226, 46, 40, 0.1)';
exports.function_danger_fill_transparent_02 = 'rgba(226, 46, 40, 0.15)';
exports.function_danger_fill_transparent_03 = 'rgba(226, 46, 40, 0.2)';
exports.function_danger_on_danger_fill = 'rgb(255, 255, 255)';
exports.function_info_100 = 'rgb(224, 233, 255)';
exports.function_info_200 = 'rgb(194, 212, 255)';
exports.function_info_300 = 'rgb(148, 180, 255)';
exports.function_info_350 = 'rgb(122, 162, 255)';
exports.function_info_400 = 'rgb(80, 131, 251)';
exports.function_info_50 = 'rgb(240, 244, 255)';
exports.function_info_500 = 'rgb(51, 109, 244)';
exports.function_info_600 = 'rgb(20, 86, 240)';
exports.function_info_700 = 'rgb(4, 66, 210)';
exports.function_info_800 = 'rgb(0, 47, 158)';
exports.function_info_900 = 'rgb(0, 34, 112)';
exports.function_info_content_default = 'rgb(20, 86, 240)';
exports.function_info_content_hover = 'rgb(51, 109, 244)';
exports.function_info_content_loading = 'rgb(148, 180, 255)';
exports.function_info_content_pressed = 'rgb(4, 66, 210)';
exports.function_info_fill_default = 'rgb(20, 86, 240)';
exports.function_info_fill_hover = 'rgb(51, 109, 244)';
exports.function_info_fill_loading = 'rgb(148, 180, 255)';
exports.function_info_fill_pressed = 'rgb(4, 66, 210)';
exports.function_info_fill_solid_01 = 'rgb(240, 244, 255)';
exports.function_info_fill_solid_02 = 'rgb(224, 233, 255)';
exports.function_info_fill_solid_03 = 'rgb(194, 212, 255)';
exports.function_info_fill_transparent_01 = 'rgba(20, 86, 240, 0.1)';
exports.function_info_fill_transparent_02 = 'rgba(20, 86, 240, 0.15)';
exports.function_info_fill_transparent_03 = 'rgba(20, 86, 240, 0.2)';
exports.function_info_on_info_fill = 'rgb(255, 255, 255)';
exports.function_success_100 = 'rgb(208, 245, 206)';
exports.function_success_200 = 'rgb(149, 229, 153)';
exports.function_success_300 = 'rgb(92, 209, 104)';
exports.function_success_350 = 'rgb(53, 189, 75)';
exports.function_success_400 = 'rgb(50, 166, 69)';
exports.function_success_50 = 'rgb(228, 250, 225)';
exports.function_success_500 = 'rgb(37, 136, 50)';
exports.function_success_600 = 'rgb(26, 117, 38)';
exports.function_success_700 = 'rgb(11, 96, 23)';
exports.function_success_800 = 'rgb(4, 67, 12)';
exports.function_success_900 = 'rgb(2, 44, 7)';
exports.function_success_content_default = 'rgb(50, 166, 69)';
exports.function_success_content_hover = 'rgb(53, 189, 75)';
exports.function_success_content_loading = 'rgb(92, 209, 104)';
exports.function_success_content_pressed = 'rgb(37, 136, 50)';
exports.function_success_fill_default = 'rgb(53, 189, 75)';
exports.function_success_fill_hover = 'rgb(92, 209, 104)';
exports.function_success_fill_loading = 'rgb(92, 209, 104)';
exports.function_success_fill_pressed = 'rgb(50, 166, 69)';
exports.function_success_fill_solid_01 = 'rgb(228, 250, 225)';
exports.function_success_fill_solid_02 = 'rgb(208, 245, 206)';
exports.function_success_fill_solid_03 = 'rgb(149, 229, 153)';
exports.function_success_fill_transparent_01 = 'rgba(53, 189, 75, 0.1)';
exports.function_success_fill_transparent_02 = 'rgba(53, 189, 75, 0.15)';
exports.function_success_fill_transparent_03 = 'rgba(53, 189, 75, 0.2)';
exports.function_success_on_success_fill = 'rgb(255, 255, 255)';
exports.function_warning_100 = 'rgb(254, 231, 205)';
exports.function_warning_200 = 'rgb(254, 196, 139)';
exports.function_warning_300 = 'rgb(255, 157, 76)';
exports.function_warning_350 = 'rgb(255, 129, 26)';
exports.function_warning_400 = 'rgb(237, 109, 12)';
exports.function_warning_50 = 'rgb(255, 243, 229)';
exports.function_warning_500 = 'rgb(194, 87, 5)';
exports.function_warning_600 = 'rgb(164, 73, 4)';
exports.function_warning_700 = 'rgb(133, 58, 5)';
exports.function_warning_800 = 'rgb(100, 43, 2)';
exports.function_warning_900 = 'rgb(59, 26, 2)';
exports.function_warning_content_default = 'rgb(237, 109, 12)';
exports.function_warning_content_hover = 'rgb(255, 129, 26)';
exports.function_warning_content_loading = 'rgb(255, 157, 76)';
exports.function_warning_content_pressed = 'rgb(194, 87, 5)';
exports.function_warning_fill_default = 'rgb(255, 129, 26)';
exports.function_warning_fill_hover = 'rgb(255, 157, 76)';
exports.function_warning_fill_loading = 'rgb(255, 157, 76)';
exports.function_warning_fill_pressed = 'rgb(237, 109, 12)';
exports.function_warning_fill_solid_01 = 'rgb(255, 243, 229)';
exports.function_warning_fill_solid_02 = 'rgb(254, 231, 205)';
exports.function_warning_fill_solid_03 = 'rgb(254, 196, 139)';
exports.function_warning_fill_transparent_01 = 'rgba(237, 109, 12, 0.1)';
exports.function_warning_fill_transparent_02 = 'rgba(237, 109, 12, 0.15)';
exports.function_warning_fill_transparent_03 = 'rgba(237, 109, 12, 0.2)';
exports.function_warning_on_warning_fill = 'rgb(255, 255, 255)';
exports.gradient_blue = 'linear-gradient(90deg, rgb(80, 131, 251) 0%, rgb(51, 109, 244) 100%)';
exports.gradient_carmine = 'linear-gradient(90deg, rgb(223, 88, 165) 0%, rgb(204, 57, 140) 100%)';
exports.gradient_green = 'linear-gradient(90deg, rgb(92, 209, 104) 0%, rgb(53, 189, 75) 100%)';
exports.gradient_indigo = 'linear-gradient(90deg, rgb(117, 125, 240) 0%, rgb(91, 101, 245) 100%)';
exports.gradient_lime = 'linear-gradient(90deg, rgb(200, 221, 95) 0%, rgb(162, 193, 11) 100%)';
exports.gradient_orange = 'linear-gradient(90deg, rgb(255, 157, 76) 0%, rgb(255, 129, 26) 100%)';
exports.gradient_purple = 'linear-gradient(90deg, rgb(159, 111, 241) 0%, rgb(141, 85, 237) 100%)';
exports.gradient_red = 'linear-gradient(90deg, rgb(255, 117, 112) 0%, rgb(245, 74, 69) 100%)';
exports.gradient_turquoise = 'linear-gradient(90deg, rgb(111, 232, 216) 0%, rgb(51, 214, 192) 100%)';
exports.gradient_violet = 'linear-gradient(90deg, rgb(222, 129, 222) 0%, rgb(207, 94, 207) 100%)';
exports.gradient_wathet = 'linear-gradient(90deg, rgb(62, 195, 247) 0%, rgb(37, 176, 231) 100%)';
exports.gradient_yellow = 'linear-gradient(90deg, rgb(255, 198, 10) 0%, rgb(217, 153, 4) 100%)';
exports.icon_disabled = 'rgb(187, 191, 196)';
exports.icon_n1 = 'rgb(43, 47, 54)';
exports.icon_n2 = 'rgb(100, 106, 115)';
exports.icon_n3 = 'rgb(143, 149, 158)';
exports.illustration_blue_a = 'rgb(0, 34, 112)';
exports.illustration_blue_b = 'rgb(0, 34, 112)';
exports.illustration_blue_c = 'rgb(0, 34, 112)';
exports.illustration_blue_d = 'rgb(0, 34, 112)';
exports.illustration_blue_e = 'rgb(51, 109, 244)';
exports.illustration_neutral_a = 'rgb(255, 255, 255)';
exports.illustration_neutral_b = 'rgb(255, 255, 255)';
exports.illustration_neutral_c = 'rgba(187, 191, 196, 0.45)';
exports.illustration_neutral_d = 'rgb(143, 149, 158)';
exports.illustration_orange = 'rgb(255, 129, 26)';
exports.illustration_red = 'rgb(245, 74, 69)';
exports.illustration_static_black = 'rgb(0, 0, 0)';
exports.illustration_static_white = 'rgb(255, 255, 255)';
exports.illustration_turquoise = 'rgb(51, 214, 192)';
exports.illustration_wathet = 'rgb(62, 195, 247)';
exports.illustration_yellow = 'rgb(255, 198, 10)';
exports.inverse_primary_content_default = 'rgb(80, 131, 251)';
exports.inverse_primary_content_hover = 'rgb(122, 162, 255)';
exports.inverse_primary_content_pressed = 'rgb(51, 109, 244)';
exports.line_border_card = 'rgb(222, 224, 227)';
exports.line_border_component = 'rgb(208, 211, 214)';
exports.line_divider_default = 'rgba(31, 35, 41, 0.15)';
exports.line_divider_module = 'rgba(31, 35, 41, 0.15)';
exports.primary_content_default = 'rgb(20, 86, 240)';
exports.primary_content_hover = 'rgb(51, 109, 244)';
exports.primary_content_loading = 'rgb(148, 180, 255)';
exports.primary_content_pressed = 'rgb(4, 66, 210)';
exports.primary_fill_default = 'rgb(20, 86, 240)';
exports.primary_fill_hover = 'rgb(51, 109, 244)';
exports.primary_fill_loading = 'rgb(148, 180, 255)';
exports.primary_fill_pressed = 'rgb(4, 66, 210)';
exports.primary_fill_solid_01 = 'rgb(240, 244, 255)';
exports.primary_fill_solid_02 = 'rgb(224, 233, 255)';
exports.primary_fill_solid_03 = 'rgb(194, 212, 255)';
exports.primary_fill_transparent_01 = 'rgba(20, 86, 240, 0.1)';
exports.primary_fill_transparent_02 = 'rgba(20, 86, 240, 0.15)';
exports.primary_fill_transparent_03 = 'rgba(20, 86, 240, 0.2)';
exports.primary_on_primary_fill = 'rgb(255, 255, 255)';
exports.primary_pri_100 = 'rgb(224, 233, 255)';
exports.primary_pri_200 = 'rgb(194, 212, 255)';
exports.primary_pri_300 = 'rgb(148, 180, 255)';
exports.primary_pri_350 = 'rgb(122, 162, 255)';
exports.primary_pri_400 = 'rgb(80, 131, 251)';
exports.primary_pri_50 = 'rgb(240, 244, 255)';
exports.primary_pri_500 = 'rgb(51, 109, 244)';
exports.primary_pri_600 = 'rgb(20, 86, 240)';
exports.primary_pri_700 = 'rgb(4, 66, 210)';
exports.primary_pri_800 = 'rgb(0, 47, 158)';
exports.primary_pri_900 = 'rgb(0, 34, 112)';
exports.shadow_default = 'rgb(31, 35, 41)';
exports.shadow_default_lg = 'rgba(31, 35, 41, 0.08)';
exports.shadow_default_md = 'rgba(31, 35, 41, 0.1)';
exports.shadow_default_sm = 'rgba(31, 35, 41, 0.12)';
exports.shadow_lg_down = '0px 6px 24px 0px rgba(31, 35, 41, 0.08)';
exports.shadow_lg_left = '-6px 0px 24px 0px rgba(31, 35, 41, 0.08)';
exports.shadow_lg_right = '6px 0px 24px 0px rgba(31, 35, 41, 0.08)';
exports.shadow_lg_up = '0px -6px 24px 0px rgba(31, 35, 41, 0.08)';
exports.shadow_md_down = '0px 4px 8px 0px rgba(31, 35, 41, 0.1)';
exports.shadow_md_left = '-4px 0px 8px 0px rgba(31, 35, 41, 0.1)';
exports.shadow_md_right = '4px 0px 8px 0px rgba(31, 35, 41, 0.1)';
exports.shadow_md_up = '0px -4px 8px 0px rgba(31, 35, 41, 0.1)';
exports.shadow_pri = 'rgb(20, 86, 240)';
exports.shadow_pri_lg = 'rgba(20, 86, 240, 0.24)';
exports.shadow_pri_lg_down = '0px 6px 24px 0px rgba(20, 86, 240, 0.24)';
exports.shadow_pri_md = 'rgba(20, 86, 240, 0.24)';
exports.shadow_pri_md_down = '0px 4px 8px 0px rgba(20, 86, 240, 0.24)';
exports.shadow_s1_down = '0px 1px 2px -2px rgba(31, 35, 41, 0.02), 0px 2px 4px 0px rgba(31, 35, 41, 0.02), 0px 2px 8px 2px rgba(31, 35, 41, 0.02)';
exports.shadow_s1_down_pri = '0px 1px 2px -2px rgba(20, 86, 240, 0.12), 0px 2px 4px 0px rgba(20, 86, 240, 0.04), 0px 2px 8px 2px rgba(20, 86, 240, 0.02)';
exports.shadow_s1_left = '-1px 0px 2px -2px rgba(31, 35, 41, 0.02), -2px 0px 4px 0px rgba(31, 35, 41, 0.02), -2px 0px 8px 2px rgba(31, 35, 41, 0.02)';
exports.shadow_s1_right = '1px 0px 2px -2px rgba(31, 35, 41, 0.02), 2px 0px 4px 0px rgba(31, 35, 41, 0.02), 2px 0px 8px 2px rgba(31, 35, 41, 0.02)';
exports.shadow_s1_up = '0px -1px 2px -2px rgba(31, 35, 41, 0.02), 0px -2px 4px 0px rgba(31, 35, 41, 0.02), 0px -2px 8px 2px rgba(31, 35, 41, 0.02)';
exports.shadow_s2_down = '0px 4px 16px 4px rgba(31, 35, 41, 0.03), 0px 4px 8px 0px rgba(31, 35, 41, 0.02), 0px 2px 4px -4px rgba(31, 35, 41, 0.02)';
exports.shadow_s2_down_pri = '0px 2px 4px -4px rgba(20, 86, 240, 0.12), 0px 4px 8px 0px rgba(20, 86, 240, 0.04), 0px 4px 16px 4px rgba(20, 86, 240, 0.03)';
exports.shadow_s2_left = '-4px 0px 16px 4px rgba(31, 35, 41, 0.03), -4px 0px 8px 0px rgba(31, 35, 41, 0.02), -2px 0px 4px -4px rgba(31, 35, 41, 0.02)';
exports.shadow_s2_right = '4px 0px 16px 4px rgba(31, 35, 41, 0.03), 4px 0px 8px 0px rgba(31, 35, 41, 0.02), 2px 0px 4px -4px rgba(31, 35, 41, 0.02)';
exports.shadow_s2_up = '0px -4px 16px 4px rgba(31, 35, 41, 0.03), 0px -4px 8px 0px rgba(31, 35, 41, 0.02), 0px -2px 4px -4px rgba(31, 35, 41, 0.02)';
exports.shadow_s3_down = '0px 6px 18px 6px rgba(31, 35, 41, 0.03), 0px 3px 6px -6px rgba(31, 35, 41, 0.05), 0px 4px 8px 0px rgba(31, 35, 41, 0.03)';
exports.shadow_s3_down_pri = '0px 3px 6px -6px rgba(20, 86, 240, 0.14), 0px 4px 8px 0px rgba(20, 86, 240, 0.06), 0px 6px 18px 6px rgba(20, 86, 240, 0.04)';
exports.shadow_s3_left = '-6px 0px 18px 6px rgba(31, 35, 41, 0.03), -3px 0px 6px -6px rgba(31, 35, 41, 0.05), -4px 0px 8px 0px rgba(31, 35, 41, 0.03)';
exports.shadow_s3_right = '6px 0px 18px 6px rgba(31, 35, 41, 0.03), 3px 0px 6px -6px rgba(31, 35, 41, 0.05), 4px 0px 8px 0px rgba(31, 35, 41, 0.03)';
exports.shadow_s3_up = '0px -6px 18px 6px rgba(31, 35, 41, 0.03), 0px -3px 6px -6px rgba(31, 35, 41, 0.05), 0px -4px 8px 0px rgba(31, 35, 41, 0.03)';
exports.shadow_s4_down = '0px 8px 24px 8px rgba(31, 35, 41, 0.04), 0px 6px 12px 0px rgba(31, 35, 41, 0.04), 0px 4px 8px -8px rgba(31, 35, 41, 0.06)';
exports.shadow_s4_down_pri = '0px 4px 8px -4px rgba(20, 86, 240, 0.16), 0px 6px 12px 0px rgba(20, 86, 240, 0.08), 0px 8px 24px 8px rgba(20, 86, 240, 0.06)';
exports.shadow_s4_left = '-8px 0px 24px 8px rgba(31, 35, 41, 0.04), -6px 0px 12px 0px rgba(31, 35, 41, 0.04), -4px 0px 8px -8px rgba(31, 35, 41, 0.06)';
exports.shadow_s4_right = '8px 0px 24px 8px rgba(31, 35, 41, 0.04), 6px 0px 12px 0px rgba(31, 35, 41, 0.04), 4px 0px 8px -8px rgba(31, 35, 41, 0.06)';
exports.shadow_s4_up = '0px -8px 24px 8px rgba(31, 35, 41, 0.04), 0px -6px 12px 0px rgba(31, 35, 41, 0.04), 0px -4px 8px -8px rgba(31, 35, 41, 0.06)';
exports.shadow_s5_down = '0px 10px 36px 10px rgba(31, 35, 41, 0.04), 0px 8px 24px 0px rgba(31, 35, 41, 0.04), 0px 6px 12px -10px rgba(31, 35, 41, 0.06)';
exports.shadow_s5_down_pri = '0px 10px 12px -10px rgba(20, 86, 240, 0.18), 0px 8px 24px 0px rgba(20, 86, 240, 0.1), 0px 10px 36px 10px rgba(20, 86, 240, 0.06)';
exports.shadow_s5_left = '-10px 0px 36px 10px rgba(31, 35, 41, 0.04), -8px 0px 24px 0px rgba(31, 35, 41, 0.04), -6px 0px 12px -10px rgba(31, 35, 41, 0.06)';
exports.shadow_s5_right = '10px 0px 36px 10px rgba(31, 35, 41, 0.04), 8px 0px 24px 0px rgba(31, 35, 41, 0.04), 6px 0px 12px -10px rgba(31, 35, 41, 0.06)';
exports.shadow_s5_up = '0px -10px 36px 10px rgba(31, 35, 41, 0.04), 0px -8px 24px 0px rgba(31, 35, 41, 0.04), 0px -6px 12px -10px rgba(31, 35, 41, 0.06)';
exports.shadow_sm_down = '0px 2px 4px 0px rgba(31, 35, 41, 0.12)';
exports.shadow_sm_left = '-2px 0px 4px 0px rgba(31, 35, 41, 0.12)';
exports.shadow_sm_right = '2px 0px 4px 0px rgba(31, 35, 41, 0.12)';
exports.shadow_sm_up = '0px -2px 4px 0px rgba(31, 35, 41, 0.12)';
exports.static_black = 'rgb(0, 0, 0)';
exports.static_black_10 = 'rgba(0, 0, 0, 0.1)';
exports.static_black_15 = 'rgba(0, 0, 0, 0.15)';
exports.static_black_20 = 'rgba(0, 0, 0, 0.2)';
exports.static_black_30 = 'rgba(0, 0, 0, 0.3)';
exports.static_black_40 = 'rgba(0, 0, 0, 0.4)';
exports.static_black_5 = 'rgba(0, 0, 0, 0.05)';
exports.static_black_50 = 'rgba(0, 0, 0, 0.5)';
exports.static_black_60 = 'rgba(0, 0, 0, 0.6)';
exports.static_black_70 = 'rgba(0, 0, 0, 0.7)';
exports.static_black_80 = 'rgba(0, 0, 0, 0.8)';
exports.static_black_90 = 'rgba(0, 0, 0, 0.9)';
exports.static_white = 'rgb(255, 255, 255)';
exports.static_white_10 = 'rgba(255, 255, 255, 0.1)';
exports.static_white_15 = 'rgba(255, 255, 255, 0.15)';
exports.static_white_20 = 'rgba(255, 255, 255, 0.2)';
exports.static_white_30 = 'rgba(255, 255, 255, 0.3)';
exports.static_white_40 = 'rgba(255, 255, 255, 0.4)';
exports.static_white_5 = 'rgba(255, 255, 255, 0.05)';
exports.static_white_50 = 'rgba(255, 255, 255, 0.5)';
exports.static_white_60 = 'rgba(255, 255, 255, 0.6)';
exports.static_white_70 = 'rgba(255, 255, 255, 0.7)';
exports.static_white_80 = 'rgba(255, 255, 255, 0.8)';
exports.static_white_90 = 'rgba(255, 255, 255, 0.9)';
exports.static_white_hover = 'rgba(255, 255, 255, 0.1)';
exports.static_white_pressed = 'rgba(255, 255, 255, 0.2)';
exports.text_caption = 'rgb(100, 106, 115)';
exports.text_disabled = 'rgb(187, 191, 196)';
exports.text_link_disabled = 'rgb(187, 191, 196)';
exports.text_link_hover = 'rgb(51, 109, 244)';
exports.text_link_loading = 'rgb(148, 180, 255)';
exports.text_link_normal = 'rgb(20, 86, 240)';
exports.text_link_pressed = 'rgb(4, 66, 210)';
exports.text_placeholder = 'rgb(143, 149, 158)';
exports.text_title = 'rgb(31, 35, 41)';
