// @universe-design/react/lib/shared/components/list/ListItem.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ListItemTitle = exports.ListItemLike = exports.ListItemDesc = exports.ListItem = exports.ListEmpty = exports.ListDivider = exports.ListCaption = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _ListCheckBoldOutlined = _interopRequireDefault(require("@universe-design/icons-react/ListCheckBoldOutlined"));
var _clsx = _interopRequireDefault(require("clsx"));
var _react = _interopRequireDefault(require("react"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _stylex2 = require("../../../stylex");
var _util = require("../../../util");
var _hooks = require("../../../config-provider/hooks");
var _textOverflow = require("../text-overflow");
require("./index.less");
const cssListItemLike = _stylex.default.compose(_stylex2.cssFlexBase, _stylex2.cssTypographyBody0);
const ListItemLike = exports.ListItemLike = /*#__PURE__*/_react.default.forwardRef((_ref, ref) => {
  let {
    className,
    ...restProps
  } = _ref;
  const cssScope = (0, _hooks.useCssScope)('list__itemLike');
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
    ref: ref,
    className: (0, _clsx.default)(cssScope(), (0, _stylex.default)(cssListItemLike), className)
  }, restProps));
});
const ListItem = exports.ListItem = /*#__PURE__*/_react.default.forwardRef((_ref2, ref) => {
  let {
    className,
    disabled = false,
    selected = false,
    active = false,
    before,
    after,
    children,
    showCheckmark = true,
    sx,
    ...restProps
  } = _ref2;
  const cssScope = (0, _hooks.useCssScope)('list__item');
  const itemLikeCssScope = (0, _hooks.useCssScope)('list__itemLike');
  return /*#__PURE__*/_react.default.createElement("li", (0, _extends2.default)({
    ref: ref,
    className: (0, _clsx.default)(className, cssScope(), itemLikeCssScope(), (0, _stylex.default)(cssListItemLike, ...(sx !== null && sx !== void 0 ? sx : [])), disabled && cssScope('-disabled'), selected && cssScope('-selected'), active && cssScope('-active'))
  }, restProps), (0, _util.isNonNullable)(before) && /*#__PURE__*/_react.default.createElement("div", {
    className: cssScope('__before')
  }, before), (0, _util.isNonNullable)(children) && /*#__PURE__*/_react.default.createElement("div", {
    className: cssScope('__content')
  }, children), (0, _util.isNonNullable)(after) ? /*#__PURE__*/_react.default.createElement("div", {
    className: cssScope('__after')
  }, after) : showCheckmark ? /*#__PURE__*/_react.default.createElement(_ListCheckBoldOutlined.default, {
    className: (0, _clsx.default)(cssScope('__checkmark'))
  }) : null);
});
const ListItemTitle = exports.ListItemTitle = /*#__PURE__*/_react.default.forwardRef((_ref3, ref) => {
  let {
    className,
    children,
    before,
    textOverflow = true,
    ...restProps
  } = _ref3;
  const cssScope = (0, _hooks.useCssScope)('list__item__title');
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
    ref: ref,
    className: (0, _clsx.default)(cssScope(), className)
  }, restProps), (0, _util.isNonNullable)(before) && /*#__PURE__*/_react.default.createElement("div", {
    className: cssScope('__before')
  }, before), (0, _util.isNonNullable)(children) && (textOverflow ? /*#__PURE__*/_react.default.createElement(_textOverflow.TextOverflow, {
    as: "div",
    lineClamp: 1
  }, children) : children));
});
const ListItemDesc = exports.ListItemDesc = /*#__PURE__*/_react.default.forwardRef((_ref4, ref) => {
  let {
    className,
    ...restProps
  } = _ref4;
  const cssScope = (0, _hooks.useCssScope)('list__item__description');
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
    ref: ref,
    className: (0, _clsx.default)(cssScope(), (0, _stylex.default)(_stylex2.cssTypographyBody2), className)
  }, restProps));
});
const ListDivider = exports.ListDivider = /*#__PURE__*/_react.default.forwardRef((_ref5, ref) => {
  let {
    className,
    fluid = false,
    marginBlock = 'both',
    ...restProps
  } = _ref5;
  const cssScope = (0, _hooks.useCssScope)('list__divider');
  return /*#__PURE__*/_react.default.createElement("hr", (0, _extends2.default)({
    ref: ref,
    className: (0, _clsx.default)(cssScope(), className, fluid && cssScope('--fluid'), cssScope(`--margin-${marginBlock}`))
  }, restProps));
});
const ListCaption = exports.ListCaption = /*#__PURE__*/_react.default.forwardRef((_ref6, ref) => {
  let {
    className,
    sx,
    ...restProps
  } = _ref6;
  const cssScope = (0, _hooks.useCssScope)('list__caption');
  const itemLikeCssScope = (0, _hooks.useCssScope)('list__itemLike');
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
    ref: ref,
    className: (0, _clsx.default)(cssScope(), itemLikeCssScope(), (0, _stylex.default)(cssListItemLike, _stylex2.cssTypographyHeadline, ...(sx || [])), className)
  }, restProps));
});
const ListEmpty = exports.ListEmpty = /*#__PURE__*/_react.default.forwardRef((_ref7, ref) => {
  let {
    className,
    ...restProps
  } = _ref7;
  const cssScope = (0, _hooks.useCssScope)('list__empty');
  const itemLikeCssScope = (0, _hooks.useCssScope)('list__itemLike');
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
    ref: ref,
    className: (0, _clsx.default)(cssScope(), itemLikeCssScope(), (0, _stylex.default)(cssListItemLike), className)
  }, restProps));
});
