// @universe-design/react/es/slider/handle.js
import clsx from 'clsx';
import React, { useEffect, useRef, useState } from 'react';
import { useCssScope } from '../config-provider/hooks';
import { Tooltip } from '../tooltip';
import { isFunction } from '../util';
import { useEventCallback } from '../hooks/useEventCallback';
export const SliderHandle = props => {
  const {
    max,
    min,
    value,
    vertical,
    dragging,
    tooltipVisible: tooltipVisibleInProps,
    tooltipPlacement,
    tipFormatter,
    getTooltipPopupContainer
  } = props;
  const cssScope = useCssScope('slider');
  const unSetVisible = typeof tooltipVisibleInProps === 'undefined';
  const [toolTipVisible, setToolTipVisible] = useState();
  const udTooltipRef = useRef(null);
  useEffect(() => {
    if (unSetVisible && typeof dragging !== 'undefined') {
      setToolTipVisible(dragging);
    } else {
      setToolTipVisible(tooltipVisibleInProps);
    }
  }, [unSetVisible, tooltipVisibleInProps, dragging]);
  useEffect(() => {
    var _udTooltipRef$current;
    // 1. 由于 `forcePopupAlign` 本身就是做了 throttle 的，所以我们用 `useEffect` 就可以了
    // 2. deps 数组先用 [value]，这种情况下如果 `min` 和 `max` 动态变化并不会触发 `forcePopupAlign`，但是大概率用户不会有拖动导致 `value` 变化的时候动态变化 `min` 和 `max` 的用法，先这样改，看看用户反馈
    (_udTooltipRef$current = udTooltipRef.current) === null || _udTooltipRef$current === void 0 ? void 0 : _udTooltipRef$current.forcePopupAlign();
  }, [value]);
  const handleTooltipVisibleChange = visible => {
    if (unSetVisible) {
      if (typeof dragging === 'undefined') {
        setToolTipVisible(visible);
      } else {
        setToolTipVisible(visible || dragging);
      }
    }
  };
  const handleHolderMouseDown = useEventCallback(ev => {
    // 解决 Safari 上拖动时光标从 pointer 变成 text 的问题
    ev.preventDefault();
  });
  return /*#__PURE__*/React.createElement(React.Fragment, null, (() => {
    const v = value ? Math.round((value - min) / (max - min) * 100) : 0;
    const style = vertical ? {
      bottom: `${v}%`
    } : {
      left: `${v}%`
    };
    const handleCls = clsx(cssScope('__handle'), {
      [cssScope('__handle--active')]: dragging
    });
    const title = isFunction(tipFormatter) ? tipFormatter(value) : `${value}`;
    if ((unSetVisible || tooltipVisibleInProps) && tipFormatter !== null) {
      return /*#__PURE__*/React.createElement(Tooltip, {
        title: title,
        visible: toolTipVisible,
        onVisibleChange: handleTooltipVisibleChange,
        getPopupContainer: getTooltipPopupContainer,
        placement: tooltipPlacement,
        udTriggerRef: udTooltipRef
      }, /*#__PURE__*/React.createElement("div", {
        style: style,
        className: cssScope('__handle-holder'),
        onMouseDown: handleHolderMouseDown
      }, /*#__PURE__*/React.createElement("div", {
        className: handleCls
      })));
    }
    return /*#__PURE__*/React.createElement("div", {
      style: style,
      className: cssScope('__handle-holder'),
      onMouseDown: handleHolderMouseDown
    }, /*#__PURE__*/React.createElement("div", {
      className: handleCls
    }));
  })());
};
