// @universe-design/react/es/slider/index.js
/* eslint-disable max-nested-callbacks */
import clsx from 'clsx';
import isEqual from 'lodash/isEqual';
import isArray from 'lodash/isArray';
import React, { useCallback, useEffect, useRef, useState } from 'react';
import { useControlledState } from '@universe-design/react-utilities';
import { useCssScope } from '../config-provider/hooks';
import { isFunction } from '../util';
import { SliderHandle } from './handle';
import { SliderMarks } from './marks';
import { SliderSteps } from './steps';
import './style';
/**
 * 单滑块 Slider 和双滑块共有的属性和方法
 * @public
 */

/**
 * 单滑块 Slider 独有的属性和方法
 * @public
 */

/**
 * 双滑块 Slider 独有的属性和方法
 * @public
 */
var DraggingStatus = /*#__PURE__*/function (DraggingStatus) {
  DraggingStatus[DraggingStatus["start"] = 0] = "start";
  DraggingStatus[DraggingStatus["dragging"] = 1] = "dragging";
  DraggingStatus[DraggingStatus["end"] = 2] = "end";
  return DraggingStatus;
}(DraggingStatus || {});
/**
 * @public
 */
export const Slider = props => {
  const {
    min = 0,
    max = 100,
    step = 1,
    dots,
    marks,
    range,
    reverse,
    vertical,
    included = true,
    disabled,
    tooltipPlacement,
    tooltipVisible: tooltipVisibleInProps,
    className,
    style,
    tipFormatter,
    getTooltipPopupContainer,
    onChange,
    onAfterChange
  } = props;
  const cssScope = useCssScope('slider');
  const cls = clsx(cssScope(), {
    [cssScope('--with-marks')]: marks,
    [cssScope('--vertical')]: vertical,
    [cssScope('--disabled')]: disabled
  }, className);
  const defaultValueInProps = props.range ? 0 : props.defaultValue || 0;
  const valueInProps = props.range ? undefined : props.value;
  const defaultRangeValue = [0, 0];
  const defaultRangeValueInProps = props.range ? props.defaultValue || defaultRangeValue : defaultRangeValue;
  const rangeValueInProps = props.range ? props.value : undefined;
  const sliderRef = useRef(null);
  const boundsRef = useRef();
  const latestValueRef = useRef();
  const [dragging, setDragging] = useState(false);
  const [draggingStatus, setDraggingStatus] = useState(DraggingStatus.end);
  const [marksBounds, setMarksBounds] = useState();
  const [marksDots, setMarksDots] = useState();
  const [prevMovedHandleIndex, setPrevMovedHandleIndex] = useState();
  const [value, setValue] = useControlledState(defaultValueInProps, valueInProps);
  const [rangeValue, setRangeValue] = useControlledState(defaultRangeValueInProps, rangeValueInProps);
  const isControlled = 'value' in props;
  const safeValue = Math.min(max, Math.max(value, min));
  const setValueWithTriggerOnChange = useCallback(data => {
    if (range && isArray(data)) {
      isFunction(onChange) && onChange(data);
      if (!isControlled) {
        setRangeValue(data);
        latestValueRef.current = data;
      }
    }
    if (!range && typeof data === 'number') {
      isFunction(onChange) && onChange(data);
      if (!isControlled) {
        setValue(data);
        latestValueRef.current = data;
      }
    }
  }, [isControlled, onChange, range, setRangeValue, setValue]);
  const getClosestHandleIndex = useCallback(data => {
    let result;
    const sortVal = [rangeValue[0], rangeValue[1]];
    const start = sortVal[0];
    const end = sortVal[1];
    if (data > end) {
      result = 1;
    } else if (data < start) {
      result = 0;
    } else {
      result = Math.abs(end - data) < Math.abs(start - data) ? 1 : 0;
    }
    return result;
  }, [rangeValue]);

  // 只计算
  const calculateValue = useCallback((x, y) => {
    var _sliderRef$current;
    const rect = (_sliderRef$current = sliderRef.current) === null || _sliderRef$current === void 0 ? void 0 : _sliderRef$current.getBoundingClientRect();
    if (!rect) {
      return undefined;
    }
    const size = vertical ? rect.height : rect.width;
    const origin = vertical ? y : rect.left;
    const startPoint = vertical ? rect.bottom : x;
    const width = Math.min(Math.max(startPoint - origin, 0), size);
    let currentValue;
    if (step) {
      if (marksDots && marksDots.length > 0) {
        const stepNumbers = [];
        for (let i = 0; i <= Math.round((max - min) / step); i++) {
          stepNumbers.push(i * step);
        }
        const mergeDots = Array.from(new Set([...stepNumbers, ...marksDots]));
        const v = Math.round(width / size * (max - min)) + min;
        const close = mergeDots.sort((a, b) => Math.abs(a - v) - Math.abs(b - v))[0];
        currentValue = close;
      } else {
        currentValue = Math.round(width / size * (max - min)) + min;
      }
    } else {
      if (marksDots && marksBounds) {
        const bound = Math.round(width / size * 100);
        marksBounds.forEach(item => {
          const start = item[0];
          const end = item[1];
          if (bound >= start && bound <= end) {
            currentValue = bound > (end - start) / 2 + start ? end : start;
          }
        });
      }
    }
    return currentValue;
  }, [vertical, marksBounds, marksDots, max, min, step]);
  const [trackStyle, setTrackStyle] = useState({});
  useEffect(() => {
    requestAnimationFrame(() => {
      if (!included && marks) {
        return;
      }
      if (!range) {
        const v = Math.round((safeValue - min) / (max - min) * 100);
        if (vertical) {
          setTrackStyle({
            bottom: 0,
            height: `${v}%`
          });
        } else {
          setTrackStyle({
            left: 0,
            width: `${v}%`
          });
        }
      } else {
        const sortVal = [rangeValue[0], rangeValue[1]];
        const start = sortVal[0];
        const end = sortVal[1];
        const s = Math.round((start - min) / (max - min) * 100);
        const w = Math.round((end - min) / (max - min) * 100) - s;
        if (vertical) {
          setTrackStyle({
            bottom: `${s}%`,
            height: `${w}%`
          });
        } else {
          setTrackStyle({
            left: `${s}%`,
            width: `${w}%`
          });
        }
      }
    });
  }, [vertical, included, marks, range, value, rangeValue, safeValue, max, min]);
  const onUp = React.useCallback(() => {
    setDraggingStatus(DraggingStatus.end);
    setDragging(false);
    boundsRef.current = undefined;
    if (draggingStatus !== DraggingStatus.start) {
      return;
    }
    if (!range) {
      onAfterChange === null || onAfterChange === void 0 ? void 0 : onAfterChange(value);
    } else {
      onAfterChange === null || onAfterChange === void 0 ? void 0 : onAfterChange(rangeValue);
    }
  }, [draggingStatus, onAfterChange, range, rangeValue, value]);
  useEffect(() => {
    document.addEventListener('pointerup', onUp);
    return () => {
      document.removeEventListener('pointerup', onUp);
    };
  }, [onUp]);
  useEffect(() => {
    const onMove = e => {
      if (disabled) {
        return;
      }
      if (draggingStatus !== DraggingStatus.start) {
        return;
      }

      // 如果检测到此时没有按键，认为已经松开
      if (e.buttons === 0) {
        onUp();
        return;
      }
      if (!dragging) {
        setDragging(true);
      }
      const data = calculateValue(e.clientX, e.clientY);
      if (!data && data !== 0) {
        return;
      }
      if (!range) {
        setValueWithTriggerOnChange(data);
      } else {
        if (typeof prevMovedHandleIndex === 'undefined' || !boundsRef.current) {
          return;
        }
        // 永远有一个点是始终保持不变点。
        const [start, end] = boundsRef.current;
        const begin = prevMovedHandleIndex === 0 ? end : start;
        const currentRangeValue = [begin, data];
        if (!isEqual(currentRangeValue, rangeValue)) {
          let [newBeginVal, newEndVal] = currentRangeValue;
          if (newBeginVal > newEndVal) {
            [newBeginVal, newEndVal] = [newEndVal, newBeginVal];
          }
          setValueWithTriggerOnChange([newBeginVal, newEndVal]);
        }
      }
    };
    document.addEventListener('pointermove', onMove);
    return () => {
      document.removeEventListener('pointermove', onMove);
    };
  }, [range, rangeValue, dragging, draggingStatus, prevMovedHandleIndex, calculateValue, setValueWithTriggerOnChange, setPrevMovedHandleIndex, disabled, onUp]);
  useEffect(() => {
    const marksNumbers = marks ? Object.keys(marks).map(parseFloat).sort((a, b) => a - b) : [];
    let numbers = marksNumbers;
    if (step && dots) {
      const stepNumbers = [];
      for (let i = 0; i <= Math.round((max - min) / step); i++) {
        stepNumbers.push(i * step);
      }
      numbers = Array.from(new Set([...stepNumbers, ...marksNumbers])).sort((a, b) => a - b);
    }
    const currentBounds = numbers.reduce((acc, cur, i) => {
      if (i === 0 && cur > min) {
        acc.push([0, cur]);
      }
      if (i === numbers.length - 1 && cur < max) {
        acc.push([cur, max]);
      }
      if (numbers[i + 1]) {
        acc.push([cur, numbers[i + 1]]);
      }
      return acc;
    }, []);
    setMarksDots(numbers);
    setMarksBounds(currentBounds);
  }, [marks, dots, step, max, min]);
  const handleMarkClick = data => {
    if (!range) {
      setValueWithTriggerOnChange(data);
    } else {
      const closestIndex = getClosestHandleIndex(data);
      setPrevMovedHandleIndex(closestIndex);
      const currentRangeValue = [rangeValue[0], rangeValue[1]];
      currentRangeValue[closestIndex] = data;
      setValueWithTriggerOnChange(currentRangeValue);
      boundsRef.current = [currentRangeValue[0], currentRangeValue[1]];
    }
  };
  return /*#__PURE__*/React.createElement("div", {
    style: style,
    className: cls,
    onPointerDown: e => {
      if (disabled) {
        return;
      }
      const target = e.target;
      const {
        classList
      } = target;
      // 点击 holder 热区的时候屏蔽一下处理
      if (!range && classList.contains(cssScope('__handle')) || classList.contains(cssScope('__handle-holder'))) {
        setDraggingStatus(DraggingStatus.start);
        return;
      }
      const data = calculateValue(e.clientX, e.clientY);
      if (!data && data !== 0) {
        return;
      }
      if (!range) {
        setValueWithTriggerOnChange(data);
      } else {
        const closestIndex = getClosestHandleIndex(data);
        // closestIndex => 0 / 1
        setPrevMovedHandleIndex(closestIndex);
        const currentRangeValue = [rangeValue[0], rangeValue[1]];
        currentRangeValue[closestIndex] = data;
        setValueWithTriggerOnChange(currentRangeValue);
        boundsRef.current = [currentRangeValue[0], currentRangeValue[1]];
      }
      setDraggingStatus(DraggingStatus.start);
    }
  }, /*#__PURE__*/React.createElement("div", {
    ref: sliderRef,
    className: cssScope('__wrapper')
  }, (() => /*#__PURE__*/React.createElement("div", {
    className: cssScope('__rail')
  }))(), (() => /*#__PURE__*/React.createElement("div", {
    className: cssScope('__track'),
    style: trackStyle
  }))(), (() => /*#__PURE__*/React.createElement(SliderSteps, {
    max: max,
    min: min,
    dots: marksDots,
    range: range,
    value: value,
    included: included,
    rangeValue: rangeValue,
    reverse: reverse,
    vertical: vertical
  }))(), (() => /*#__PURE__*/React.createElement(SliderMarks, {
    max: max,
    min: min,
    marks: marks,
    reverse: reverse,
    vertical: vertical,
    onMarkClick: handleMarkClick
  }))(), (() => {
    const placement = vertical ? tooltipPlacement || 'right' : tooltipPlacement;
    if (!range) {
      return /*#__PURE__*/React.createElement(SliderHandle, {
        max: max,
        min: min,
        value: safeValue,
        vertical: vertical,
        dragging: dragging,
        tooltipVisible: tooltipVisibleInProps,
        tooltipPlacement: placement,
        tipFormatter: tipFormatter
      });
    }
    const [start, end] = rangeValue;
    const pi = prevMovedHandleIndex;
    const startDragging = dragging && (pi === 0 && boundsRef.current && end === boundsRef.current[1] || boundsRef.current && end === boundsRef.current[0]);
    const endDragging = dragging && (pi === 1 && boundsRef.current && start === boundsRef.current[0] || boundsRef.current && start === boundsRef.current[1]);
    return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(SliderHandle, {
      max: max,
      min: min,
      value: start,
      vertical: vertical,
      dragging: startDragging,
      tooltipVisible: tooltipVisibleInProps,
      tooltipPlacement: placement,
      tipFormatter: tipFormatter,
      getTooltipPopupContainer: getTooltipPopupContainer
    }), /*#__PURE__*/React.createElement(SliderHandle, {
      max: max,
      min: min,
      value: end,
      vertical: vertical,
      dragging: endDragging,
      tooltipVisible: tooltipVisibleInProps,
      tooltipPlacement: placement,
      tipFormatter: tipFormatter,
      getTooltipPopupContainer: getTooltipPopupContainer
    }));
  })()));
};
