// @universe-design/react/es/slider/marks.js
import clsx from 'clsx';
import React, { isValidElement } from 'react';
import { useCssScope } from '../config-provider/hooks';
import { Text } from '../text';
export const SliderMarks = props => {
  const {
    max,
    min,
    marks,
    reverse,
    vertical,
    onMarkClick
  } = props;
  const cssScope = useCssScope('slider');
  const markCls = clsx(cssScope('__marks'), {
    [cssScope('__marks--vertical')]: vertical
  });
  const handleMarkClick = (e, value) => {
    e.stopPropagation();
    onMarkClick(value);
  };
  return marks ? /*#__PURE__*/React.createElement("div", {
    className: markCls
  }, (() => {
    if (marks) {
      const keys = Object.keys(marks);
      return keys.map(parseFloat).sort((a, b) => a - b).map(key => {
        const item = marks[key];
        const isObject = typeof item === 'object' && ! /*#__PURE__*/isValidElement(item);
        const label = isObject ? item.label : item;
        const style = isObject ? item.style : undefined;
        if (!label && label !== 0) {
          return null;
        }
        const cls = clsx(cssScope('__marks-item'));
        const point = key;
        const mergedStyle = {
          ...style,
          ...(vertical ? {
            [reverse ? 'top' : 'bottom']: `${(point - min) / (max - min) * 100}%`
          } : {
            [reverse ? 'right' : 'left']: `${(point - min) / (max - min) * 100}%`
          })
        };
        return /*#__PURE__*/React.createElement(Text, {
          as: "div",
          variant: "body2",
          key: key,
          style: mergedStyle,
          className: cls,
          onMouseDown: e => handleMarkClick(e, point)
        }, label);
      });
    }
    return null;
  })()) : null;
};
