// @universe-design/react/src/dialog/Confirm.tsx
import type { ReactNode } from 'react';
import React, { useEffect, useState } from 'react';
import ErrorFilled from '@universe-design/icons-react/ErrorFilled';
import InfoFilled from '@universe-design/icons-react/InfoFilled';
import SucceedFilled from '@universe-design/icons-react/SucceedFilled';
import WarningFilled from '@universe-design/icons-react/WarningFilled';
import clsx from 'clsx';
import stylex from '@universe-design/stylex';
import { cssTypographyBody0, cssTypographyTitle4 } from '../stylex';
import { sxDialogIconWrapper, sxDialogIcon, sxModalContent } from './sx';
import { Dialog } from './dialog';
import { Button } from '../button';
import { useCssScope, useLocale } from '../config-provider/hooks';
import type { StyleProps } from '../util';
import type { StylexProps } from '@universe-design/stylex/react';
import { isFunction, isNonNullable } from '../util';
import type { ModalProps } from './Modal';
import { useEffectOnce } from '../hooks';
import { FirstLineAlign } from '../shared/components/FirstLineAlign';

import './style';

interface ConfirmContentProps
  extends Pick<
      ModalProps,
      'title' | 'okText' | 'okButtonProps' | 'cancelText' | 'cancelButtonProps' | 'bodyStyle' | 'extraAction'
    >,
    StyleProps,
    StylexProps {
  /**
   * 自定义图标
   */
  icon?: ReactNode;
  /**
   * 内容
   */
  content?: ReactNode;
  /**
   * 是否显示取消按钮
   * @defaultValue true
   */
  enableCancelButton?: boolean;
  /**
   * 指定自动获得焦点的按钮
   * @defaultValue `ok`
   */
  autoFocusButton?: null | 'ok' | 'cancel';
  /**
   * 确认框类型，会影响图标类型和颜色
   */
  type?: 'info' | 'success' | 'error' | 'warning' | 'confirm';
  /**
   * @internal
   */
  okButtonRef?: React.Ref<HTMLButtonElement>;
  /**
   * @internal
   */
  cancelButtonRef?: React.Ref<HTMLButtonElement>;
  /**
   * @internal
   */
  onClose?: () => void;
  /**
   * 取消回调，参数为关闭函数，返回 `Promise` 时 `resolve` 后自动关闭
   */
  onCancel?: (close?: () => void) => void | Promise<unknown>;
  /**
   * 点击确定回调，参数为关闭函数，返回 `Promise` 时 `resolve` 后自动关闭
   */
  onOk?: (close?: () => void) => void | Promise<unknown>;
}

/**
 * @public
 * @docgen
 */
export interface ConfirmProps
  extends Pick<
      ModalProps,
      // from inner dialog
      | 'zIndex'
      | 'centered'
      | 'visible'
      | 'getContainer'
      | 'keyboardClosable'
      | 'destroyOnClose'
      | 'forceRender'
      | 'onAfterClose'
      | 'onAfterOpen'
      | 'safeMargin'
      | 'enableAnimation'
      | 'initialFocusRef'
      | 'finalFocusRef'
      | 'enableFocusLock'
      | 'enableRestoreFocus'
      | 'width'
      | 'minWidth'
      | 'maxWidth'
      // from modal
      | 'mask'
      | 'style'
      | 'maskStyle'
      | 'className'
      | 'wrapClassName'
      | 'maskClosable'
    >,
    Omit<ConfirmContentProps, 'className' | 'style' | 'sx'> {}

export const ConfirmContent = (props: ConfirmContentProps) => {
  const {
    icon: iconProp,
    title,
    content,
    cancelText,
    okText,
    cancelButtonProps,
    okButtonProps,
    enableCancelButton = true,
    autoFocusButton = 'ok',
    bodyStyle,
    onCancel,
    onOk,
    onClose,
    type: typeProp,
    extraAction,
    // style
    style,
    className,
    // ref
    okButtonRef,
    cancelButtonRef,
    sx = [],
  } = props;

  const type = typeProp ?? 'confirm';

  const icon = (() => {
    /**
     * 现实效果: 用户传入icon: null时 icon不显示 (当用户不穿入icon属性时 icon为undefined)
     * 使用isNonNullable函数 -> iconProp为 null 与 undefined 时均根据typeProp选择相应icon
     * 此处对用户不传iconProp 和 传入iconProp: null分开处理
     */
    if (iconProp !== undefined) {
      return iconProp;
    }

    switch (typeProp) {
      case 'error':
        return <ErrorFilled />;
      case 'success':
        return <SucceedFilled />;
      case 'info':
        return <InfoFilled />;
      case 'confirm':
      case 'warning':
        return <WarningFilled />;
      default:
        return null;
    }
  })();

  const cssScope = useCssScope('confirm');
  const locale = useLocale();
  const cancelRef = React.useRef<HTMLButtonElement>(null);
  React.useImperativeHandle<HTMLButtonElement | null, HTMLButtonElement | null>(
    cancelButtonRef,
    () => cancelRef.current,
  );
  const okRef = React.useRef<HTMLButtonElement>(null);
  React.useImperativeHandle<HTMLButtonElement | null, HTMLButtonElement | null>(okButtonRef, () => okRef.current);
  const [okLoading, setOkLoading] = useState<boolean>(Boolean(okButtonProps?.loading));
  const [cancelLoading, setCancelLoading] = useState<boolean>(Boolean(cancelButtonProps?.loading));

  const triggerClose = () => onClose && onClose();

  // 自动设置焦点，回车后会触发相关事件
  useEffectOnce(() => {
    let timerId: number;
    if (autoFocusButton) {
      timerId = window.setTimeout(() => {
        autoFocusButton === 'cancel' ? cancelRef.current?.focus() : okRef.current?.focus();
      });
    }
    return () => {
      if (timerId) {
        window.clearTimeout(timerId);
      }
    };
  });

  useEffect(() => {
    setOkLoading(Boolean(okButtonProps?.loading));
    setCancelLoading(Boolean(cancelButtonProps?.loading));
  }, [okButtonProps, cancelButtonProps]);

  const handleOnConfirmClick = (event: React.MouseEvent<HTMLElement, MouseEvent>) => {
    if (okButtonProps && okButtonProps.onClick) {
      okButtonProps.onClick(event);
    }

    if (!onOk) {
      triggerClose();
    } else {
      if (onOk.length) {
        onOk(triggerClose);
      } else {
        const result = onOk();
        if (result && result.then) {
          setOkLoading(true);
          result.then(
            () => {
              setOkLoading(false);
              triggerClose();
            },
            (e: Error) => {
              console.error(e);
              setOkLoading(false);
            },
          );
        } else {
          triggerClose();
        }
      }
    }
  };

  const handleOnCancelClick = (event: React.MouseEvent<HTMLElement, MouseEvent>) => {
    if (cancelButtonProps && cancelButtonProps.onClick) {
      cancelButtonProps.onClick(event);
    }

    if (!onCancel) {
      triggerClose();
    } else {
      if (onCancel.length) {
        onCancel(triggerClose);
      } else {
        const result = onCancel();
        if (result && result.then) {
          setCancelLoading(true);
          result.then(
            () => {
              setCancelLoading(false);
              triggerClose();
            },
            (e: Error) => {
              console.error(e);
              setCancelLoading(false);
            },
          );
        } else {
          triggerClose();
        }
      }
    }
  };

  const headerNode = (
    <div className={cssScope('__header')}>
      <FirstLineAlign
        className={cssScope('__title')}
        sx={[cssTypographyTitle4]}
        beforeProps={
          isNonNullable(icon)
            ? {
                className: cssScope('__icon'),
                sx: [sxDialogIconWrapper],
                children: <div className={stylex(sxDialogIcon)}>{icon}</div>,
              }
            : undefined
        }
      >
        <div className={cssScope('__titleContent')}>{title}</div>
      </FirstLineAlign>
      {isNonNullable(content) ? (
        <div style={bodyStyle} className={clsx(cssScope('__body'), stylex(cssTypographyBody0))}>
          {content}
        </div>
      ) : null}
    </div>
  );

  const renderFooter = () => (
    <div className={cssScope('__footer')}>
      {extraAction ? <div className={cssScope('__footer__extraAction')}>{extraAction}</div> : null}
      <div className={cssScope('__footer__btns')}>
        <Button ref={okRef} type="primary" {...okButtonProps} onClick={handleOnConfirmClick} loading={okLoading}>
          {okText || locale.components.modal.okText}
        </Button>
        {enableCancelButton && (
          <Button ref={cancelRef} {...cancelButtonProps} loading={cancelLoading} onClick={handleOnCancelClick}>
            {cancelText || locale.components.modal.cancelText}
          </Button>
        )}
      </div>
    </div>
  );

  return (
    <div
      className={clsx(
        className,
        cssScope('__content'),
        {
          [cssScope('-hasIcon')]: icon,
          [cssScope(`-${type}`)]: true,
        },
        stylex(sxModalContent.root, cssTypographyBody0, ...sx),
      )}
      style={style}
    >
      {headerNode}
      {renderFooter()}
    </div>
  );
};

// eslint-disable-next-line max-lines-per-function
export const Confirm = (props: ConfirmProps) => {
  const {
    icon,
    title,
    content,
    keyboardClosable = true,
    mask = true,
    centered = true,
    destroyOnClose,
    visible,
    cancelText,
    okText,
    cancelButtonProps,
    okButtonProps,
    enableCancelButton = true,
    maskStyle,
    bodyStyle,
    autoFocusButton = 'ok',
    onCancel,
    onOk,
    onClose, // 内部属性
    onAfterOpen,
    onAfterClose,
    className,
    style,
    wrapClassName,
    type = 'confirm',
    extraAction,
    maskClosable = false,
    ...restProps
  } = props;

  const cssScope = useCssScope('confirm');

  const triggerClose = () => onClose && onClose();

  const handleOnClose = () => {
    triggerClose();
    isFunction(onCancel) && onCancel();
  };

  const handleOnAfterClose = () => {
    isFunction(onAfterClose) && onAfterClose();
  };

  const initialFocusRef = React.useRef<HTMLButtonElement>(null);

  return (
    <Dialog
      visible={visible}
      centered={centered}
      onClose={handleOnClose}
      overlay={mask}
      onAfterOpen={onAfterOpen}
      onAfterClose={handleOnAfterClose}
      destroyOnClose={destroyOnClose}
      overlayStyle={maskStyle}
      wrapperClassName={wrapClassName}
      contentStyle={style}
      overlayClosable={maskClosable}
      keyboardClosable={keyboardClosable}
      contentClassName={clsx(cssScope(), className)}
      size="small"
      initialFocusRef={autoFocusButton ? initialFocusRef : undefined}
      {...restProps}
    >
      <ConfirmContent
        {...{
          icon,
          title,
          content,
          cancelText,
          okText,
          cancelButtonProps,
          okButtonProps,
          enableCancelButton,
          autoFocusButton,
          bodyStyle,
          onCancel,
          onOk,
          onClose,
          type,
          extraAction,
          sx: [
            centered &&
              stylex.create({
                margin: 'auto 0',
              }),
          ],
          okButtonRef: autoFocusButton === 'ok' ? initialFocusRef : undefined,
          cancelButtonRef: autoFocusButton === 'cancel' ? initialFocusRef : undefined,
        }}
      />
    </Dialog>
  );
};

export function withWarn(props: ConfirmProps): ConfirmProps {
  return {
    enableCancelButton: false,
    ...props,
    type: 'warning',
  };
}

export function withInfo(props: ConfirmProps): ConfirmProps {
  return {
    enableCancelButton: false,
    ...props,
    type: 'info',
  };
}

export function withSuccess(props: ConfirmProps): ConfirmProps {
  return {
    enableCancelButton: false,
    ...props,
    type: 'success',
  };
}

export function withError(props: ConfirmProps): ConfirmProps {
  return {
    enableCancelButton: false,
    ...props,
    type: 'error',
  };
}

export function withConfirm(props: ConfirmProps): ConfirmProps {
  return {
    enableCancelButton: true,
    ...props,
    type: 'confirm',
  };
}
