// @universe-design/react/src/dialog/index.tsx
import React from 'react';
import ReactDOM from 'react-dom';
import type { ConfirmProps } from './Confirm';
import { Confirm, withConfirm, withError, withInfo, withSuccess, withWarn } from './Confirm';
import type { ModalProps, ModalContentProps } from './Modal';
import { Modal as ModalComponents, ModalContent } from './Modal';
import { isFunction } from '../util';
import { CrossRootConfigProvider } from '../config-provider';
import type { DialogIsntance } from './hooks/RefDialog';

type Destroy = () => void;
const destroyMap = new Map<string, Destroy>();

type StaticFn = (config: ConfirmProps) => DialogIsntance<ConfirmProps>;

type ModalStaticFn = {
  info: StaticFn;
  success: StaticFn;
  error: StaticFn;
  warning: StaticFn;
  confirm: StaticFn;
};

type ModalType = typeof ModalComponents & ModalStaticFn & { destroyAll: Destroy; Content: typeof ModalContent };

export const Dialog = ModalComponents as ModalType;

function confirm({ onAfterClose, visible, ...config }: ConfirmProps) {
  const uuid = `ud-confirm-instance-${new Date().getTime()}`;
  const container = document.createElement('div');
  document.body.appendChild(container);

  let currentConfig: ConfirmProps;

  function destroy() {
    const unmountResult = ReactDOM.unmountComponentAtNode(container);
    if (unmountResult && container.parentNode) {
      container.parentNode.removeChild(container);
    }
    // config.onCancel && config.onCancel();
    destroyMap.delete(uuid);
  }

  function render(renderConfig: ConfirmProps) {
    ReactDOM.render(
      <CrossRootConfigProvider>
        <Confirm {...renderConfig} />
      </CrossRootConfigProvider>,
      container,
    );
  }

  function close() {
    currentConfig = {
      ...currentConfig,
      visible: false,
    };
    render(currentConfig);
  }

  function update(newConfig: ConfirmProps | ((cfg: ConfirmProps) => ConfirmProps)) {
    if (isFunction(newConfig)) {
      currentConfig = newConfig(currentConfig);
    } else {
      currentConfig = {
        ...currentConfig,
        ...newConfig,
      };
    }
    render(currentConfig);
  }

  currentConfig = {
    ...config,
    onAfterClose: () => {
      isFunction(onAfterClose) && onAfterClose();
      destroy();
    },
    visible: true,
    onClose: close,
  };

  render(currentConfig);

  destroyMap.set(uuid, destroy);

  return {
    destroy: close,
    update,
  };
}

Dialog.destroyAll = () => {
  destroyMap.forEach(fn => fn());
  destroyMap.clear();
};

Dialog.Content = ModalContent;

Dialog.confirm = (config: ConfirmProps) => confirm(withConfirm(config));

Dialog.info = (config: ConfirmProps) => confirm(withInfo(config));

Dialog.success = (config: ConfirmProps) => confirm(withSuccess(config));

Dialog.error = (config: ConfirmProps) => confirm(withError(config));

Dialog.warning = (config: ConfirmProps) => confirm(withWarn(config));

export type { ConfirmProps, ModalProps, ModalContentProps };

export { ConfirmContent, Confirm } from './Confirm';

export { useDialog } from './hooks/useDialog';
export type { DialogIsntance } from './hooks/RefDialog';
export { DialogCancelType, DialogLayoutMode } from './dialog';
export type { DialogLayoutModeProps } from './dialog';
