// @universe-design/react/src/dialog/Modal.tsx
/* eslint-disable @typescript-eslint/no-use-before-define */
import type { CSSProperties, ReactNode, HTMLAttributes } from 'react';
import React, { useCallback, useEffect, useState, forwardRef } from 'react';
import clsx from 'clsx';
import stylex from '@universe-design/stylex';
import { cssTypographyTitle4, cssTypographyBody0 } from '../stylex';
import type { DialogProps } from './dialog';
import { Dialog, DialogCancelType, useDialogLayoutMode } from './dialog';
import type { ButtonProps } from '../button';
import { Button } from '../button';
import CloseOutlined from '@universe-design/icons-react/CloseOutlined';
import { useCssScope, useLocale } from '../config-provider/hooks';
import { useSize } from '../hooks';
import { useResizeObserver, useCloneElementRef } from '@universe-design/react-utilities';
import type { StylexProps } from '@universe-design/stylex/react';
import { isFunction, isNonNullable } from '../util';
import type { ThunkElement } from '../util/getThunkElement';
import { getThunkElement } from '../util/getThunkElement';
import { FirstLineAlign } from '../shared/components/FirstLineAlign';
import { sxModalContent } from './sx';

import './style';

export interface ModalContentExoticProps {
  /**
   * 对话框头部
   */
  header?: ReactNode;
  /**
   * 标题
   */
  title?: ReactNode;
  /**
   * 说明
   */
  description?: ReactNode;
  /**
   * 是否显示右上角的关闭按钮
   * @defaultValue true
   */
  closable?: boolean;
  /**
   * 是否显示取消按钮
   * @defaultValue true
   */
  enableCancelButton?: boolean;
  /**
   * 取消按钮文字
   * @defaultValue `取消`
   */
  cancelText?: string;
  /**
   * 取消按钮组件属性
   */
  cancelButtonProps?: ButtonProps;
  /**
   * 确认按钮文字
   * @defaultValue `确定`
   */
  okText?: string;
  /**
   * 确认按钮组件属性组件属性
   */
  okButtonProps?: ButtonProps;
  /**
   * 对话框主体 `style` 属性透传
   */
  bodyStyle?: CSSProperties;
  /**
   * 确定按钮是否为加载状态
   * @defaultValue false
   */
  confirmLoading?: boolean;
  /**
   * 底部内容，当不需要默认底部按钮时，可以设为 `footer={null}`
   */
  footer?: ReactNode;
  /**
   * 点击遮罩层、右上角叉、取消按钮或按下ESC键的回调
   */
  onCancel?: (
    e: React.MouseEvent<HTMLElement, MouseEvent> | React.KeyboardEvent | undefined,
    payload: { cancelType: DialogCancelType },
  ) => void;
  /**
   * 点击确定回调
   */
  onOk?: (e: React.MouseEvent<HTMLElement, MouseEvent>) => void;
  /**
   * 额外操作
   */
  extraAction?: ReactNode;
  /**
   * body 区域的高度自适应模式
   * - `off` 可视区高度按内容高度撑开
   * - `scroll` 可视区高度由内容和安全边距决定，内容可滚动
   * - `stretch` 可视区域内容一样高，内容拉伸，用户需要提供内容拉伸能力
   * @defaultValue `scroll`
   */
  bodyHeightResponsiveMode?: 'off' | 'scroll' | 'stretch';
  /**
   * 内容的最大高度
   */
  maxHeight?: React.CSSProperties['maxHeight'];
}

function useCheckVerticalOverflow<T extends HTMLElement>({ target }: { target: ThunkElement<T> }) {
  const [overflowTop, setOverflowTop] = useState(false);
  const [overflowBottom, setOverflowBottom] = useState(false);

  const checkVerticalOverflow = useCallback(() => {
    const targetEl = getThunkElement(target);
    if (!targetEl) {
      return;
    }

    const { scrollTop, scrollHeight, clientHeight } = targetEl;
    const targetInset = {
      top: 0,
      bottom: clientHeight,
    };
    const childInset = {
      top: -scrollTop,
      bottom: -scrollTop + scrollHeight,
    };

    if (targetInset.top > childInset.top) {
      setOverflowTop(true);
    } else {
      setOverflowTop(false);
    }

    if (targetInset.bottom < childInset.bottom) {
      setOverflowBottom(true);
    } else {
      setOverflowBottom(false);
    }
  }, [target]);

  return {
    overflow: {
      top: overflowTop,
      bottom: overflowBottom,
    },
    checkVerticalOverflow,
  };
}

/**
 * @public
 * @docgen
 * <Dialog.Content /> 的属性。
 */
export interface ModalContentProps
  extends ModalContentExoticProps,
    Omit<HTMLAttributes<HTMLDivElement>, 'title'>,
    StylexProps {}

/**
 * @public
 * @docgen
 * <Dialog /> 的属性。
 */
export interface ModalProps
  extends Pick<
      DialogProps,
      | 'zIndex'
      | 'centered'
      | 'visible'
      | 'getContainer'
      | 'keyboardClosable'
      | 'destroyOnClose'
      | 'forceRender'
      | 'onAfterClose'
      | 'onAfterOpen'
      | 'size'
      | 'safeMargin'
      | 'enableAnimation'
      | 'initialFocusRef'
      | 'finalFocusRef'
      | 'enableFocusLock'
      | 'enableRestoreFocus'
      | 'width'
      | 'minWidth'
      | 'maxWidth'
    >,
    ModalContentExoticProps,
    Omit<HTMLAttributes<HTMLDivElement>, 'title'> {
  /**
   * 是否展示遮罩
   * @defaultValue true
   */
  mask?: boolean;
  /**
   * 点击遮罩是否允许关闭
   * @defaultValue true
   */
  maskClosable?: boolean;
  /**
   * 对话框 `style` 属性透传
   */
  style?: CSSProperties;
  /**
   * 对话框遮罩 `style` 属性透传
   */
  maskStyle?: CSSProperties;
  /**
   * 对话框 `className` 属性透传
   */
  className?: string;
  /**
   * 对话框外包裹 `className` 属性透传
   */
  wrapClassName?: string;
  transitionName?: string;
  maskTransitionName?: string;
  /**
   * @internal
   */
  onClose?: () => void;
}

export const ModalContent = forwardRef<HTMLDivElement, ModalContentProps>((props, ref) => {
  const cssScope = useCssScope('modal');

  const {
    header,
    title,
    description,
    closable = true,
    enableCancelButton = true,
    children,
    cancelText,
    okText,
    cancelButtonProps,
    okButtonProps,
    bodyStyle = {},
    confirmLoading,
    footer,
    onCancel,
    onOk,
    extraAction,
    style,
    className,
    sx = [],
    bodyHeightResponsiveMode = 'scroll',
    maxHeight,
    ...restProps
  } = props;

  const locale = useLocale();
  const layoutMode = useDialogLayoutMode();

  const handleCloseIcon = (e: React.MouseEvent<HTMLElement, MouseEvent>) => {
    isFunction(onCancel) && onCancel(e, { cancelType: DialogCancelType.CloseIcon });
  };

  const closeIconNode = closable ? (
    <Button className={cssScope('__close')} onClick={handleCloseIcon} icon={<CloseOutlined />} type="icon" />
  ) : null;

  const bodyScrollable = bodyHeightResponsiveMode === 'scroll';
  const bodyStretch = bodyHeightResponsiveMode === 'stretch';
  const [headerSizeRef, headerSize] = useSize<HTMLDivElement>({
    resizeObserverDisabled: !bodyScrollable && !bodyStretch,
  });
  const [footerSizeRef, footerSize] = useSize<HTMLDivElement>({
    resizeObserverDisabled: !bodyScrollable && !bodyStretch,
  });

  const contentCls = clsx(
    className,
    cssScope('__content'),
    closeIconNode && cssScope('-closable'),
    stylex(
      sxModalContent.root,
      bodyScrollable && sxModalContent.scrollable,
      cssTypographyBody0,
      bodyHeightResponsiveMode === 'stretch' && sxModalContent.stretch,
      ...sx,
    ),
  );

  const handleCancel = (e: React.MouseEvent<HTMLElement, MouseEvent>) => {
    if (cancelButtonProps && cancelButtonProps.onClick) {
      cancelButtonProps.onClick(e);
    }
    isFunction(onCancel) && onCancel(e, { cancelType: DialogCancelType.CancelButton });
  };

  const handleOk = (e: React.MouseEvent<HTMLElement, MouseEvent>) => {
    if (okButtonProps && okButtonProps.onClick) {
      okButtonProps.onClick(e);
    }
    isFunction(onOk) && onOk(e);
  };

  /**
   * Title ReactNode
   */
  let headerNode: ReactNode = null;
  if (header !== null) {
    headerNode = (
      <div ref={headerSizeRef} className={cssScope('__header')}>
        {isNonNullable(header) ? (
          header
        ) : (
          <FirstLineAlign className={cssScope('__title')} sx={[cssTypographyTitle4]}>
            <div className={cssScope('__titleContent')}>{title}</div>
          </FirstLineAlign>
        )}
        {isNonNullable(description) ? (
          <div className={clsx(cssScope('__des'), stylex(cssTypographyBody0))}>{description}</div>
        ) : null}
      </div>
    );
  }

  /**
   * Footer ReactNode
   */
  const footerNode = (() => {
    const footerContentNode = (
      <>
        {extraAction ? <div className={cssScope('__footer__extraAction')}>{extraAction}</div> : null}
        <div className={cssScope('__footer__btns')}>
          <Button
            type="primary"
            {...okButtonProps}
            loading={confirmLoading || okButtonProps?.loading}
            onClick={handleOk}
          >
            {okText || locale.components.modal.okText}
          </Button>
          {enableCancelButton && (
            <Button {...cancelButtonProps} onClick={handleCancel}>
              {cancelText || locale.components.modal.cancelText}
            </Button>
          )}
        </div>
      </>
    );

    if (footer) {
      return (
        <div ref={footerSizeRef} className={cssScope('__footer')}>
          {footer}
        </div>
      );
    }
    if (footer === null) {
      return null;
    }
    return (
      <div ref={footerSizeRef} className={cssScope('__footer')}>
        {footerContentNode}
      </div>
    );
  })();

  /**
   * Body ReactNode
   */

  // body 内容超出增加分割线
  const bodyRef = useResizeObserver<HTMLDivElement>(() => {
    checkVerticalOverflow();
  });
  const { overflow: bodyOverflow, checkVerticalOverflow } = useCheckVerticalOverflow({
    target: bodyRef,
  });
  const childrenRef = useResizeObserver<HTMLDivElement>(() => {
    checkVerticalOverflow();
  });
  useEffect(() => {
    const body = bodyRef.current;
    if (!body) {
      return;
    }
    body.addEventListener('scroll', checkVerticalOverflow);
    return () => {
      body.removeEventListener('scroll', checkVerticalOverflow);
    };
  }, [bodyRef, checkVerticalOverflow]);

  const childrenMaybeWithRef = useCloneElementRef({
    children,
    ref: childrenRef,
  });

  const { height: bodyHeight } = bodyStyle;
  const bodyMinHeight = (() => {
    if (isNonNullable(bodyStyle.minHeight)) {
      return bodyStyle.minHeight;
    }
    if (layoutMode === 'legacy') {
      return 0;
    }
    // 62 是 body 的最小高度，一个字段 32 + 8 + 22
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    return 32 + 8 + 22;
  })();

  const bodyNode = isNonNullable(children)
    ? React.createElement(
        'div',
        {
          style: {
            ...bodyStyle,
            minHeight: bodyMinHeight,
            overflowY: bodyScrollable ? 'auto' : undefined,
          },
          className: clsx(
            cssScope('__body'),
            headerNode && bodyOverflow.top && cssScope('__body--top-overflow'),
            footerNode && bodyOverflow.bottom && cssScope('__body--bottom-overflow'),
            stylex(bodyStretch && stylex.create({ overflow: 'hidden' })),
          ),
          ref: bodyRef,
        },
        childrenMaybeWithRef,
      )
    : null;

  const contentStyle = (() => {
    // 24 目前在 body 底部的 margin
    const bodyMarginBottom = 24;
    const sharedStyle = { ...style, maxHeight };
    if (bodyScrollable || bodyStretch) {
      const bodyMinHeightStr = (() => {
        if (!bodyNode) {
          return;
        }
        if (isNonNullable(bodyHeight)) {
          return typeof bodyHeight === 'number' ? `${bodyHeight}px` : bodyHeight;
        }
        return typeof bodyMinHeight === 'number' ? `${bodyMinHeight}px` : bodyMinHeight;
      })();
      const minHeight = `calc(${[
        `${headerSize.height + footerSize.height + (bodyNode ? bodyMarginBottom : 0)}px`,
        bodyMinHeightStr,
      ]
        .filter(Boolean)
        .join(' + ')})`;
      return { ...sharedStyle, minHeight };
    }
    return sharedStyle;
  })();

  return (
    <div className={contentCls} style={contentStyle} ref={ref} {...restProps}>
      {headerNode}
      {closeIconNode}
      {bodyNode}
      {footerNode}
    </div>
  );
});

// eslint-disable-next-line max-lines-per-function
export const Modal = forwardRef<HTMLDivElement, ModalProps>((props, ref) => {
  const {
    title,
    zIndex,
    description,
    closable = true,
    enableCancelButton = true,
    maskClosable = true,
    keyboardClosable = true,
    centered = true,
    width,
    visible,
    children,
    cancelText,
    okText,
    cancelButtonProps,
    okButtonProps,
    mask = true,
    style,
    maskStyle,
    bodyStyle,
    enableAnimation,
    transitionName,
    maskTransitionName,
    confirmLoading,
    // SSR compatiable
    getContainer,
    footer,
    onCancel,
    onOk,
    onAfterOpen,
    onAfterClose,
    className,
    wrapClassName,
    extraAction,
    size = 'medium',
    header,
    safeMargin,
    onClose,
    bodyHeightResponsiveMode,
    maxHeight,
    ...restProps
  } = props;

  const cssScope = useCssScope('modal');
  const layoutMode = useDialogLayoutMode();

  const handleOnClose = (
    e: React.MouseEvent<HTMLElement, MouseEvent> | React.KeyboardEvent | undefined,
    payload: { cancelType: DialogCancelType },
  ) => {
    isFunction(onCancel) && onCancel(e, payload);
    if (e?.isDefaultPrevented()) {
      return;
    }
    // 这个逻辑只给 imperative dialog 使用
    isFunction(onClose) && onClose();
  };

  const shouldShowShadow = !mask;

  return (
    <Dialog
      ref={ref}
      width={width}
      zIndex={zIndex}
      visible={visible}
      centered={centered}
      overlay={mask}
      overlayStyle={maskStyle}
      wrapperClassName={wrapClassName}
      overlayClosable={maskClosable}
      contentStyle={style}
      contentClassName={clsx(cssScope(), className)}
      keyboardClosable={keyboardClosable}
      enableAnimation={enableAnimation}
      overlayTransitionClassName={transitionName}
      contentTransitionClassName={maskTransitionName}
      getContainer={getContainer}
      onClose={handleOnClose}
      onAfterOpen={onAfterOpen}
      onAfterClose={onAfterClose}
      size={size}
      safeMargin={safeMargin}
      {...restProps}
    >
      <ModalContent
        {...{
          header,
          title,
          description,
          closable,
          enableCancelButton,
          children,
          cancelText,
          okText,
          cancelButtonProps,
          okButtonProps,
          bodyStyle,
          confirmLoading,
          footer,
          onCancel,
          onOk,
          extraAction,
          sx: [
            layoutMode === 'legacy'
              ? []
              : [
                  centered &&
                    stylex.create({
                      margin: 'auto 0',
                    }),
                  shouldShowShadow &&
                    stylex.create({
                      boxShadow: $token['shadow-s5-down'],
                    }),
                ],
          ],
          bodyHeightResponsiveMode,
          maxHeight,
        }}
      />
    </Dialog>
  );
});
