// @universe-design/react/src/table/hooks/usePagination.tsx
import { useState } from 'react';
import type { PaginationProps } from '../../pagination/Pagination';
import type { TablePaginationConfig } from '../interface';
import { DEFAULT_CURRENT, DEFAULT_PAGE_SIZE } from '../constants';

export default function usePagination(
  listTotal: number,
  pagination: TablePaginationConfig | false | undefined,
  onPaginationChange: (params: { current: number; pageSize: number; total: number }) => void,
): [TablePaginationConfig, (() => void) | null, () => TablePaginationConfig] {
  const { total: paginationTotal = 0, ...paginationObj } = typeof pagination === 'object' ? pagination : {};

  const [innerPagination, setInnerPagination] = useState({
    current: paginationObj.defaultCurrent || DEFAULT_CURRENT,
    pageSize: paginationObj.pageSize || paginationObj.defaultPageSize || DEFAULT_PAGE_SIZE,
  });

  if (pagination === false) {
    return [{}, null, () => ({})];
  }

  const mergedPagination = {
    ...innerPagination,
    ...paginationObj,
    total: paginationTotal > 0 ? paginationTotal : listTotal,
  };

  if (!paginationTotal) {
    // 用户未设置 total 时，检查页数不合法 重置到1
    const maxPage = Math.ceil(listTotal / mergedPagination.pageSize);
    if (maxPage < mergedPagination.current) {
      mergedPagination.current = 1;
    }
  }

  const setPagination = (current = 1, pageSize?: number) => {
    setInnerPagination({
      current,
      pageSize: pageSize || mergedPagination.pageSize,
    });
  };

  const onInternalChange: PaginationProps['onChange'] = (current, pageSize) => {
    const paginationPageSize = mergedPagination.pageSize;
    if (pageSize && pageSize !== paginationPageSize) {
      current = 1;
    }
    if (paginationObj.onChange) {
      paginationObj.onChange(current, pageSize);
    }

    setPagination(current, pageSize);
    onPaginationChange({
      current,
      pageSize: pageSize || paginationPageSize,
      total: mergedPagination.total,
    });
  };

  const getPaginationParam = () => {
    const param = {
      ...mergedPagination,
    };
    return param;
  };

  return [
    {
      ...mergedPagination,
      onChange: onInternalChange,
    },
    setPagination,
    getPaginationParam,
  ];
}
