// @universe-design/react/src/table/hooks/useSorter.tsx
import React, { useCallback, useMemo, useState } from 'react';
import clsx from 'clsx';
import { Tooltip } from '../../tooltip';
import ExpandDownFilled from '@universe-design/icons-react/ExpandDownFilled';
import ExpandUpFilled from '@universe-design/icons-react/ExpandUpFilled';
import SortZToAOutlined from '@universe-design/icons-react/SortZToAOutlined';
import SortAToZOutlined from '@universe-design/icons-react/SortAToZOutlined';
import { getColumnKey, getColumnPos, renderColumnTitle } from '../utils';
import type {
  TransformColumns,
  ColumnsType,
  Key,
  ColumnType,
  SortOrder,
  ColumnTitleProps,
  SorterResult,
  ColumnGroupType,
  TableLocale,
  CompareFn,
} from '../interface';
import { TableSort } from '../interface';
import { IconButton } from '../../button';

function getMultiplePriority<RecordType>(column: ColumnType<RecordType>): number | false {
  if (typeof column.sorter === 'object' && typeof column.sorter.multiple === 'number') {
    return column.sorter.multiple;
  }
  return false;
}

function getSortFunction<RecordType>(sorter: ColumnType<RecordType>['sorter']): CompareFn<RecordType> | false {
  if (typeof sorter === 'function') {
    return sorter;
  }
  if (sorter && typeof sorter === 'object' && sorter.compare) {
    return sorter.compare;
  }
  return false;
}

function nextSortDirection(sortDirections: SortOrder[], current: SortOrder | null) {
  if (!current) {
    return sortDirections[0];
  }
  return sortDirections[sortDirections.indexOf(current) + 1];
}

export interface SortState<RecordType> {
  column: ColumnType<RecordType>;
  key: Key;
  sortOrder: SortOrder | null;
  multiplePriority: number | false;
}

function collectSortStates<RecordType>(
  columns: ColumnsType<RecordType>,
  init: boolean,
  pos?: string,
): SortState<RecordType>[] {
  let sortStates: SortState<RecordType>[] = [];

  function pushState(column: ColumnsType<RecordType>[number], columnPos: string) {
    sortStates.push({
      column,
      key: getColumnKey(column, columnPos),
      multiplePriority: getMultiplePriority(column),
      sortOrder: column.sortOrder ?? null,
    });
  }

  (columns || []).forEach((column, index) => {
    const columnPos = getColumnPos(index, pos);

    if ((column as ColumnGroupType<RecordType>).children) {
      if ('sortOrder' in column) {
        // Controlled
        pushState(column, columnPos);
      }
      sortStates = [
        ...sortStates,
        ...collectSortStates((column as ColumnGroupType<RecordType>).children, init, columnPos),
      ];
    } else if (column.sorter) {
      if ('sortOrder' in column) {
        // Controlled
        pushState(column, columnPos);
      } else if (init && column.defaultSortOrder) {
        // Default sorter
        sortStates.push({
          column,
          key: getColumnKey(column, columnPos),
          multiplePriority: getMultiplePriority(column),
          sortOrder: column.defaultSortOrder,
        });
      }
    }
  });

  return sortStates;
}

function stateToInfo<RecordType>(sorterStates: SortState<RecordType>) {
  const { column, sortOrder } = sorterStates;
  return {
    column,
    order: sortOrder,
    field: column.dataIndex,
    columnKey: column.key,
  };
}

function generateSorterInfo<RecordType>(
  sorterStates: SortState<RecordType>[],
): SorterResult<RecordType> | SorterResult<RecordType>[] {
  const list = sorterStates.filter(({ sortOrder }) => sortOrder).map(stateToInfo);

  if (list.length <= 1) {
    return list[0] || {};
  }

  return list;
}

export function getSortData<RecordType>(
  data: readonly RecordType[],
  sortStates: SortState<RecordType>[],
  childrenColumnName: string,
): RecordType[] {
  const innerSorterStates = sortStates
    .slice()
    .sort((a, b) => (b.multiplePriority as number) - (a.multiplePriority as number));

  const cloneData = data.slice();

  const runningSorters = innerSorterStates.filter(
    ({ column: { sorter }, sortOrder }) => getSortFunction(sorter) && sortOrder,
  );

  // Skip if no sorter needed
  if (!runningSorters.length) {
    return cloneData;
  }

  return cloneData
    .sort((record1, record2) => {
      for (let i = 0; i < runningSorters.length; i += 1) {
        const sorterState = runningSorters[i];
        const {
          column: { sorter },
          sortOrder,
        } = sorterState;

        const compareFn = getSortFunction(sorter);

        if (compareFn && sortOrder) {
          const compareResult = compareFn(record1, record2, sortOrder);

          if (compareResult !== 0) {
            return sortOrder === TableSort.ASCEND ? compareResult : -compareResult;
          }
        }
      }

      return 0;
    })
    .map<RecordType>(record => {
      const subRecords = (record as any)[childrenColumnName];
      if (subRecords) {
        return {
          ...record,
          [childrenColumnName]: getSortData(subRecords, sortStates, childrenColumnName),
        };
      }
      return record;
    });
}

interface SorterConfig<RecordType> {
  prefix: string;
  mergedColumns: ColumnsType<RecordType>;
  onSorterChange: (
    sorterResult: SorterResult<RecordType> | SorterResult<RecordType>[],
    sortStates: SortState<RecordType>[],
  ) => void;
  sortDirections: SortOrder[];
  tableLocale?: TableLocale;
  showSorterTooltip?: boolean;
}

export default function useSorter<RecordType>({
  prefix,
  mergedColumns,
  onSorterChange,
  sortDirections: tableSortDirections,
  tableLocale,
  showSorterTooltip: tableShowSorterTooltip,
}: SorterConfig<RecordType>): [
  TransformColumns<RecordType>,
  SortState<RecordType>[],
  ColumnTitleProps<RecordType>,
  () => SorterResult<RecordType> | SorterResult<RecordType>[],
] {
  const [sortStates, setSortStates] = useState<SortState<RecordType>[]>(collectSortStates(mergedColumns, true));

  const mergedSorterStates = useMemo(() => {
    let validate = true;
    const collectedStates = collectSortStates(mergedColumns, false);

    if (!collectedStates.length) {
      return sortStates;
    }

    const validateStates: SortState<RecordType>[] = [];

    function patchStates(state: SortState<RecordType>) {
      if (validate) {
        validateStates.push(state);
      } else {
        validateStates.push({
          ...state,
          sortOrder: null,
        });
      }
    }

    let multipleMode: boolean | null = null;
    collectedStates.forEach(state => {
      if (multipleMode === null) {
        patchStates(state);

        if (state.sortOrder) {
          if (state.multiplePriority === false) {
            validate = false;
          } else {
            multipleMode = true;
          }
        }
      } else if (multipleMode && state.multiplePriority !== false) {
        patchStates(state);
      } else {
        validate = false;
        patchStates(state);
      }
    });

    return validateStates;
  }, [mergedColumns, sortStates]);

  // Get render columns title required props
  const columnTitleSorterProps = useMemo<ColumnTitleProps<RecordType>>(() => {
    const sortColumns = mergedSorterStates.map(({ column, sortOrder }) => ({
      column,
      order: sortOrder,
    }));

    return {
      sortColumns,
      sortColumn: sortColumns[0]?.column,
      sortOrder: sortColumns[0]?.order,
    };
  }, [mergedSorterStates]);

  const triggerSorter = useCallback(
    (sortState: SortState<RecordType>) => {
      let newSorterStates;

      if (
        sortState.multiplePriority === false ||
        !mergedSorterStates.length ||
        mergedSorterStates[0].multiplePriority === false
      ) {
        newSorterStates = [sortState];
      } else {
        newSorterStates = [...mergedSorterStates.filter(({ key }) => key !== sortState.key), sortState];
      }

      setSortStates(newSorterStates);
      onSorterChange(generateSorterInfo(newSorterStates), newSorterStates);
    },
    [mergedSorterStates, onSorterChange],
  );

  const transformColumns = useCallback(
    (innerColumns: ColumnsType<RecordType>) => {
      const injectSorter = (
        columns: ColumnsType<RecordType>,
        sorterSates: SortState<RecordType>[],
        pos?: string,
      ): ColumnsType<RecordType> =>
        columns.map((column, index) => {
          const columnPos = getColumnPos(index, pos);
          let newColumn: ColumnsType<RecordType>[number] = column;

          if (newColumn.sorter) {
            const sortDirections: SortOrder[] = newColumn.sortDirections || tableSortDirections;
            const showSorterTooltip = newColumn.showSorterTooltip ?? tableShowSorterTooltip;
            const columnKey = getColumnKey(newColumn, columnPos);
            const sorterState = sorterSates.find(({ key }) => key === columnKey);
            const sorterOrder = sorterState ? sorterState.sortOrder : null;
            const nextSortOrder = nextSortDirection(sortDirections, sorterOrder);
            let upNode: React.ReactNode = null;
            let downNode: React.ReactNode = null;

            if (sortDirections.includes(TableSort.ASCEND) && sortDirections.includes(TableSort.DESCEND)) {
              upNode = (
                <ExpandUpFilled
                  className={clsx(`${prefix}-column-sorter-up`, {
                    active: sorterOrder === TableSort.ASCEND,
                  })}
                />
              );

              downNode = (
                <ExpandDownFilled
                  className={clsx(`${prefix}-column-sorter-down`, {
                    active: sorterOrder === TableSort.DESCEND,
                  })}
                />
              );
            } else {
              if (sortDirections.includes(TableSort.ASCEND)) {
                upNode = (
                  <IconButton
                    color={sorterOrder === TableSort.ASCEND ? 'primary' : undefined}
                    size="small"
                    selected={sorterOrder === TableSort.ASCEND}
                    // info={sorterOrder === TableSort.ASCEND}
                    icon={<SortAToZOutlined />}
                  />
                );
              }
              if (sortDirections.includes(TableSort.DESCEND)) {
                downNode = (
                  <IconButton
                    color={sorterOrder === TableSort.DESCEND ? 'primary' : undefined}
                    size="small"
                    selected={sorterOrder === TableSort.DESCEND}
                    // info={sorterOrder === TableSort.DESCEND}
                    icon={<SortZToAOutlined />}
                  />
                );
              }
            }

            const { cancelSort, triggerAsc, triggerDesc } = tableLocale || {};
            let sortTip: string | undefined = cancelSort;
            if (nextSortOrder === TableSort.DESCEND) {
              sortTip = triggerDesc;
            } else if (nextSortOrder === TableSort.ASCEND) {
              sortTip = triggerAsc;
            }
            newColumn = {
              ...newColumn,
              className: clsx(newColumn.className, {
                [`${prefix}-column-sort`]: sorterOrder,
              }),
              title: (renderProps: ColumnTitleProps<RecordType>) => {
                const renderSortTitle = (
                  <>
                    {renderColumnTitle(column.title, renderProps)}
                    <Tooltip title={sortTip} disabled={!showSorterTooltip}>
                      <span
                        className={clsx(`${prefix}-column-sorter`, {
                          [`${prefix}-column-sorter-full`]: upNode && downNode,
                        })}
                      >
                        <span className={`${prefix}-column-sorter-inner`}>
                          {upNode}
                          {downNode}
                        </span>
                      </span>
                    </Tooltip>
                  </>
                );
                return renderSortTitle;
              },
              onHeaderCell: col => {
                const cell: React.HTMLAttributes<HTMLElement> = (column.onHeaderCell && column.onHeaderCell(col)) || {};
                const originOnClick = cell.onClick;
                cell.onClick = (event: React.MouseEvent<HTMLElement>) => {
                  triggerSorter({
                    column,
                    key: columnKey,
                    sortOrder: nextSortOrder,
                    multiplePriority: getMultiplePriority(column),
                  });

                  if (originOnClick) {
                    originOnClick(event);
                  }
                };

                cell.className = clsx(cell.className, `${prefix}-column-has-sorters`);

                return cell;
              },
            };
          }

          if ('children' in newColumn) {
            newColumn = {
              ...newColumn,
              children: injectSorter(newColumn.children, sorterSates, columnPos),
            };
          }

          return newColumn;
        });

      return injectSorter(innerColumns, mergedSorterStates);
    },
    [mergedSorterStates, tableSortDirections, tableShowSorterTooltip, prefix, tableLocale, triggerSorter],
  );

  const getSorters = useCallback(() => generateSorterInfo(mergedSorterStates), [mergedSorterStates]);

  return [transformColumns, mergedSorterStates, columnTitleSorterProps, getSorters];
}
