// @universe-design/react/src/input/components/Counter.tsx
import clsx from 'clsx';
import React from 'react';
import { useCssScope } from '../../config-provider';
import type { TagProps } from '../../tag';
import { Tag } from '../../tag';

interface CounterProps {
  showCount?: boolean;
  count?: number;
  maxLength?: number;
  disabled?: boolean;
  prefix?: string;
  tagSize?: TagProps['size'];
  style?: React.CSSProperties;
}
export default function Counter(props: CounterProps) {
  const { showCount, count, maxLength, disabled, tagSize = 'extra-small' } = props;
  const cssScope = useCssScope('input');
  if (!showCount) {
    return null;
  }
  const hasExceed = count && maxLength && maxLength < count;
  return (
    <Tag
      size={tagSize}
      className={cssScope('__count')}
      color={disabled ? 'neutral-option' : 'neutral-solid'}
      disabled={disabled}
      style={props.style}
    >
      <span
        className={clsx(hasExceed && cssScope('__count-exceed'), count && cssScope('__count-current'))}
        data-testid="current-length"
      >
        {count}
      </span>
      {maxLength && <span>/{maxLength}</span>}
    </Tag>
  );
}
