// @universe-design/react/src/tooltip/Tooltip.tsx
import type { CSSProperties, MouseEventHandler, ReactNode } from 'react';
import React, { cloneElement, forwardRef, isValidElement, useMemo } from 'react';
import clsx from 'clsx';
import { useCssScope } from '../config-provider/hooks';
import { usePopperSidecar } from '../config-provider/hooks/usePopperSidecar';
import type { UDTriggerProps, UDAlignType } from '../shared/components/trigger';
import type { PositionType } from '../util';
import type { PopperProps, PopperRef } from '../shared/components/popper';

import './style';

export type TooltipPlacement = Exclude<PositionType, 'topCenter' | 'leftCenter' | 'rightCenter' | 'bottomCenter'>;

export const splitObject = (obj: any, keys: string[]) => {
  const picked: any = {};
  const omitted: any = { ...obj };
  keys.forEach(key => {
    if (obj && key in obj) {
      picked[key] = obj[key];
      delete omitted[key];
    }
  });
  return { picked, omitted };
};

export interface BaseTooltipProps
  extends Pick<
      PopperProps,
      | 'prefixCls'
      | 'className'
      | 'openClassName'
      | 'showAction'
      | 'hideAction'
      | 'visible'
      | 'defaultVisible'
      | 'onVisibleChange'
      | 'mouseEnterDelay'
      | 'mouseLeaveDelay'
      | 'destroyOnHide'
      | 'disabled'
      | 'getPopupContainer'
      | 'checkOverflow'
      | 'udTriggerRef'
      | 'trigger'
      // new popper props
      | 'offset'
      | 'popupPortalContainer'
      | 'flip'
      | 'shift'
      | 'enableAutoHide'
      | 'eventListeners'
      | 'enableMaxHeight'
      | 'middlewares'
      | 'inline'
      | 'boundary'
      | 'initialFocus'
      | 'enableFocusLock'
      | 'enableRestoreFocus'
      | 'boundaryPadding'
      | 'popupStopPropagation'
    >,
    Pick<UDTriggerProps, 'popupAlign'> {
  /**
   * 是否显示指向小箭头
   * @deprecated 已废弃，请使用 `showArrow`
   * @defaultValue false
   */
  hideArrow?: boolean;
  /**
   * 是否显示指向小箭头
   * @defaultValue true
   */
  showArrow?: boolean;
  /**
   * 最大宽度
   * @deprecated 请使用 overlayStyle 进行覆盖
   */
  maxWidth?: string | number;
  /**
   * 气泡框位置
   * @defaultValue `top`
   */
  placement?: TooltipPlacement;
  /**
   * 卡片外容器 `style` 属性透传
   */
  overlayStyle?: CSSProperties;
  /**
   * 卡片外容器 `className` 属性透传
   */
  overlayClassName?: string;
  /**
   * @deprecated 边界溢出检查设置，已废弃，直接使用 `popupAlign.overflow` 进行规则覆盖
   */
  adjustOverflow?: boolean | UDAlignType['overflow'];
  children?: ReactNode;
  /**
   * 卡片外容器的鼠标移入事件
   */
  onOverlayMouseEnter?: MouseEventHandler<HTMLDivElement>;
  /**
   * 卡片外容器的鼠标移出事件
   */
  onOverlayMouseLeave?: MouseEventHandler<HTMLDivElement>;
  /**
   * 是否开启动画
   * @defaultValue true
   */
  enableAnimation?: boolean;

  /**
   * 是否给子元素添加wrapper，解决disabled态表单元素无法触发的问题
   * @defaultValue false
   */
  wrapDisabledChildren?: boolean;
}
/**
 * @public
 * @docgen
 */
export type TooltipProps = BaseTooltipProps &
  ((
    | {
        /**
         * 提示文字
         */
        title: ReactNode;
        overlay?: undefined;
      }
    | { title?: undefined; overlay: ReactNode }
  ) & {
    /**
     * @internal
     */
    arrowSize?: {
      x: number;
      y: number;
    };
    /**
     * @internal
     */
    arrowHasStroke?: boolean;
  });

export const Tooltip = forwardRef<PopperRef, TooltipProps>((props, ref) => {
  const {
    prefixCls,
    title,
    overlay,
    trigger = 'hover',
    // maxWidth, // max-width 现在加在了 popup 上可以直接用 overlayStyle 去覆盖
    placement = 'top',
    overlayStyle,
    overlayClassName,
    hideArrow: hideArrowProp = false,
    showArrow: showArrowProp = true,
    // adjustOverflow = true,
    children,
    onOverlayMouseEnter,
    onOverlayMouseLeave,
    mouseLeaveDelay = 0,
    wrapDisabledChildren = false,
    arrowHasStroke = false,
    ...restProps
  } = props;
  const cssScope = useCssScope('tooltip', prefixCls);
  const Popper = usePopperSidecar();
  const showArrow = 'hideArrow' in props ? !hideArrowProp : showArrowProp;

  const getDisabledCompatibleChildren = React.useCallback(
    (element: React.ReactElement<any>) => {
      if (!element.props.disabled) {
        return element;
      }
      // Pick some layout related style properties up to span
      // Prevent layout bugs
      const { picked, omitted } = splitObject(element.props.style, [
        'position',
        'left',
        'right',
        'top',
        'bottom',
        'float',
        'display',
        'zIndex',
      ]);
      const spanStyle = {
        display: 'inline-block', // default inline-block is important
        ...picked,
        cursor: 'not-allowed',
        width: element.props.block ? '100%' : null,
      };
      const buttonStyle = {
        ...omitted,
        pointerEvents: 'none',
      };
      const child = cloneElement(element, {
        style: buttonStyle,
        className: null,
      });
      return (
        <span
          style={spanStyle}
          className={clsx(element.props.className, `${cssScope()}-disabled-compatible-wrapper`)}
          data-testid="tooltip-disable-wrapper"
        >
          {child}
        </span>
      );
    },
    [cssScope],
  );

  const mergedChildren = useMemo(() => {
    if (!isValidElement(children)) {
      return (
        <span className={cssScope('__invalid-children-wrapper')} data-testid="invalid-children-wrapper">
          {children}
        </span>
      );
    }

    /**
     * if children is disabled Button or native disabled button,
     * wrap children with <span>,
     * because disabled button won't fire mouseenter/mouseleave/mouseover events for some browsers
     */
    if (wrapDisabledChildren) {
      return getDisabledCompatibleChildren(children);
    }
    return children;
  }, [children, cssScope, getDisabledCompatibleChildren, wrapDisabledChildren]);

  const popup = React.useMemo(
    () => (
      <div
        onMouseEnter={onOverlayMouseEnter}
        onMouseLeave={onOverlayMouseLeave}
        className={cssScope('-content')}
        role="tooltip"
      >
        {title ?? overlay}
      </div>
    ),
    [cssScope, onOverlayMouseEnter, onOverlayMouseLeave, overlay, title],
  );
  if (!(title ?? overlay)) {
    return <>{children}</>;
  }
  return (
    <Popper
      ref={ref}
      {...{
        prefixCls: cssScope(),
        popupPlacement: placement,
        action: trigger,
        popup,
        children: mergedChildren,
        // arrow
        showArrow,
        arrowHasStroke,
        // popupMotion
        popupMotion: 'tooltip-zoom',
        popupStyle: overlayStyle,
        popupClassName: overlayClassName,
        mouseLeaveDelay,
        defaultMainAxisOffsetWithoutArrow: 4,
      }}
      {...restProps}
      __screenshotAlwaysOpenEnabled
    />
  );
});

if (__DEV__) {
  Tooltip.displayName = 'Tooltip';
}
