// @byted-service/bytedtrace/lib/reporter/reporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AgentReporter = void 0;
const tslib_1 = require("tslib");
/**
 * 上报 transaction
 */
const status = tslib_1.__importStar(require("../status"));
const gConsts = tslib_1.__importStar(require("../consts"));
const gTypes = tslib_1.__importStar(require("../types"));
const provider_1 = require("./provider");
const serializer_1 = require("../serializer");
// Reporter's default flush interval
const DEFAULT_FLUSH_INTERVAL_MS = 1000; // 1s
class AgentReporter {
    static fixSerializerOption(options) {
        if (!options) {
            options = {
                compression: gConsts.HEADER_COMPRESSION,
                contentCompression: gConsts.CONTENT_COMPRESSION,
                bigEndian: gConsts.DEFAULT_BIG_ENDIAN,
                useDic: gConsts.DEFAULT_USE_DIC,
            };
        }
        if (typeof options.version !== 'number' || options.version <= 0) {
            options.version = gConsts.GLOBAL_TRACE_VERSION;
        }
        if (typeof options.protoType !== 'number' || options.protoType <= 0) {
            options.protoType = gConsts.DEFAULT_PROTO_TYPE;
        }
        if (typeof options.reserved !== 'number' || options.reserved <= 0) {
            options.reserved = gConsts.RESERVED;
        }
        return options;
    }
    static getDefaultAgentReporterOption(options = {}) {
        if (typeof options.queueLen !== 'number' || options.queueLen <= 0) {
            options.queueLen = gConsts.DEFAULT_QUEUE_SIZE;
        }
        if (typeof options.providerQueueLen !== 'number' ||
            options.providerQueueLen <= 0) {
            options.providerQueueLen =
                gConsts.DEFAULT_AGENT_PROVIDER_QUEUE_SIZE;
        }
        if (typeof options.flushIntervalMs !== 'number' ||
            options.flushIntervalMs <= 0) {
            options.flushIntervalMs = DEFAULT_FLUSH_INTERVAL_MS;
        }
        if (typeof options.maxDataLen !== 'number' || options.maxDataLen <= 0) {
            options.maxDataLen = gConsts.DEFAULT_MAX_BAG_SIZE;
        }
        options.serializerOp = AgentReporter.fixSerializerOption(options.serializerOp);
        return options;
    }
    constructor(options, env, statusCounter) {
        this.lastFlushTimeMs = 0;
        this.options = options;
        this.env = env;
        this.statusCounter = statusCounter;
        this.provider = new provider_1.AgentProvider({
            queueLen: this.options.providerQueueLen,
            statusCounter: this.statusCounter,
            logger: this.options.logger,
        });
        this.queue = [];
        this.options.serializerOp.logger = options.logger;
        this.flowSerializer = new serializer_1.FlowSerializer(this.options.serializerOp, env);
        this.run();
    }
    /**
     * 如果 queue 中长度大于 queueLen 怎丢弃
     * @param txn
     */
    report(txn) {
        if (this.queue.length < this.options.queueLen) {
            this.queue.push(txn);
        }
        if (this.status === gTypes.ReporterStatus.wait) {
            this.run();
        }
        // 没有启动 则启动定时 flush
        if (!this.flushIntervalId) {
            this.flushWithInterval();
        }
    }
    close() {
        clearInterval(this.flushIntervalId);
        this.status = gTypes.ReporterStatus.close;
        this.provider.close();
    }
    run() {
        while (true) {
            if (this.status === gTypes.ReporterStatus.close) {
                // 主动关闭需要将所有数据 append 和做最后一次 flush
                while (this.queue.length > 0) {
                    this.append(this.queue.shift());
                }
                this.flush();
                return;
            }
            if (this.queue.length > 0) {
                this.append(this.queue.shift());
            }
            else {
                // 小于等于 0 进入 wait 状态
                this.status = gTypes.ReporterStatus.wait;
                return;
            }
        }
    }
    flushWithInterval() {
        this.flushIntervalId = setInterval(() => {
            if (Date.now() - this.lastFlushTimeMs >
                this.options.flushIntervalMs) {
                this.flush();
            }
        }, this.options.flushIntervalMs).unref();
    }
    flush() {
        let data;
        try {
            data = this.flowSerializer.getSerializedData();
        }
        catch (error) /* istanbul ignore next */ {
            this.statusCounter.recordWarningWithRateLimit(false, status.ErrSerializeFail, 'serialize failed, error: %o', error);
        }
        /* istanbul ignore if */
        if (!data) {
            return;
        }
        try {
            this.provider.write(data);
            this.statusCounter
                .inc(status.CounterTxnReportBps, data.length, true)
                .catch((error) => /* istanbul ignore next */ {
                this.statusCounter.recordWarningWithRateLimit(false, status.ErrIncTxnReportBpsCounterFail, 'inc reporter bps counter fail error: %o', error);
            });
        }
        catch (error) /* istanbul ignore next */ {
            this.statusCounter.recordWarningWithRateLimit(false, status.ErrProviderFlushFail, 'provider flush failed, error: %o', error);
        }
        this.lastFlushTimeMs = Date.now();
    }
    append(txn) {
        try {
            this.flowSerializer.append(txn);
            const length = this.flowSerializer.size();
            if (length > this.options.maxDataLen) {
                this.flush();
            }
            this.statusCounter
                .inc(status.CounterTxnReportQps, 1, true)
                .catch((error) => {
                /* istanbul ignore next */
                this.statusCounter.recordWarningWithRateLimit(false, status.ErrIncTxnReportQpsCounterFail, 'inc reporter qps counter fail %o', error);
            });
            this.statusCounter
                .inc(status.CounterSpanReportQps, txn.getSpanCount(), true)
                .catch((error) => {
                /* istanbul ignore next */
                this.statusCounter.recordWarningWithRateLimit(false, status.ErrIncSpanReportQpsCounterFail, 'Inc reporter span qps counter fail %o', error);
            });
        }
        catch (error) {
            if (error.message === gConsts.TXN_LEN_OVER_FLOW) {
                this.statusCounter.recordWarningWithRateLimit(false, status.ErrTxnDataLenOverflow, 'transaction data len overflow!');
            }
            else {
                /* istanbul ignore next */
                this.statusCounter.recordWarningWithRateLimit(false, status.ErrSerializeAppendFail, 'serializer append failed! error: %o', error);
            }
        }
    }
}
exports.AgentReporter = AgentReporter;
