// @byted-service/bytedtrace/lib/event/logEvent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LogEvent = void 0;
const consts_1 = require("../consts");
const event_1 = require("./event");
class LogEvent {
    /**
     * 返回输出 log level 为 info 的 TraceEvent 实例
     * @param name
     * @returns
     */
    static createInfoLogEvent(name) {
        return LogEvent._createLogEvent(name, consts_1.LogLevels.LEVEL_INFO, false);
    }
    /**
     * 返回输出 log level 为 warn 的 TraceEvent 实例
     * @param name
     * @returns
     */
    static createWarnLogEvent(name) {
        return LogEvent._createLogEvent(name, consts_1.LogLevels.LEVEL_WARN, true);
    }
    /**
     * 返回输出 log level 为 error 的 TraceEvent 实例
     * @param name
     * @returns
     */
    static createErrorLogEvent(name) {
        return LogEvent._createLogEvent(name, consts_1.LogLevels.LEVEL_ERROR, true);
    }
    /**
     * 返回输出 log level 为 fatal 的 TraceEvent 实例
     * @param name
     * @returns
     */
    static createFatalLogEvent(name) {
        return LogEvent._createLogEvent(name, consts_1.LogLevels.LEVEL_FATAL, true);
    }
    /**
     * 返回输出 log level 为 LogLevels.LEVEL_NO_DEFINE 的 TraceEvent 实例
     * @param eventType
     * @param name
     * @returns
     */
    static createLogEvent(eventType, name) {
        return new event_1.TraceEvent({
            tags: {},
            eventType,
            name,
            logLevel: consts_1.LogLevels.LEVEL_NO_DEFINE,
        });
    }
    static _createLogEvent(name, logLevel, emitMetrics) {
        const event = new event_1.TraceEvent({
            tags: {},
            eventType: consts_1.LOG_EVENT_TYPE,
            name,
            logLevel,
        });
        event.setEmitMetrics(emitMetrics).setEmitLog(true);
        return event;
    }
}
exports.LogEvent = LogEvent;
