// @universe-design/react/src/divider/index.tsx
import type { FC } from 'react';
import React from 'react';
import clsx from 'clsx';
import stylex from '@universe-design/stylex';
import { cssBlockBase, cssTypographyTitle4, cssTypographyBody0 } from '../stylex';
import { useCssScope } from '../config-provider/hooks';

import { sxDashed, sxDirection, sxDivider, sxTitle, sxWithText, sxLeft, sxRight, sxWithTextDashed } from './sx';

export interface DividerProps {
  /**
   * `style` 属性透传
   */
  style?: React.CSSProperties;
  /**
   * `className` 属性透传
   */
  className?: string;
  /**
   * 分割线文字内容
   */
  text?: string;
  /**
   * 分割线方向
   * @defaultValue `"horizontal"`
   */
  type?: 'horizontal' | 'vertical';
  /**
   * 分割线标题的位置
   * @defaultValue `"center"`
   */
  orientation?: 'left' | 'right' | 'center';
  /**
   * 是否虚线
   * @defaultValue `false`
   */
  dashed?: boolean;
  /**
   * 文字是否显示为普通正文样式
   * @defaultValue `false`
   */
  plain?: boolean;
  /**
   * 分割线文字内容，同 `text`
   */
  children?: React.ReactNode;
}

export const Divider: FC<DividerProps> = props => {
  const {
    text,
    children,
    style,
    className,
    type = 'horizontal',
    plain = false,
    dashed = false,
    orientation = 'center',
  } = props;

  const cssScope = useCssScope('divider');
  const hasText = Boolean(children) || Boolean(text);

  const cssTypo = (() => {
    if (!hasText || plain) {
      return cssTypographyBody0;
    }
    return cssTypographyTitle4;
  })();

  const cls = clsx(
    cssScope(),
    cssScope(`--${type}`),
    {
      [cssScope('--with-text')]: hasText,
      [cssScope(`--with-text-${orientation}`)]: hasText,
      [cssScope('--dashed')]: dashed,
      [cssScope('--plain')]: plain,
    },
    stylex(
      sxDivider,
      cssBlockBase,
      cssTypo,
      sxDirection(type),
      hasText && sxWithText,
      dashed && sxDashed,
      orientation === 'left' && sxLeft,
      orientation === 'right' && sxRight,
      hasText && dashed && sxWithTextDashed,
    ),
    className,
  );

  const renderTitle = () => {
    if (hasText) {
      return <span className={clsx(cssScope('__title'), stylex(sxTitle))}>{text ?? children}</span>;
    }
    return null;
  };

  return (
    <div style={style} className={cls}>
      {renderTitle()}
    </div>
  );
};
