// @universe-design/react/src/divider/sx.ts
import stylex from '@universe-design/stylex';

export const sxDivider = stylex.create({
  borderTop: `1px solid ${$token['line-divider-default']}`,
});

export const sxWithText = stylex.create({
  borderTopWidth: 0,
  whiteSpace: 'nowrap',
  textAlign: 'center',
  '::before': {
    position: 'relative',
    top: '50%',
    width: '50%',
    borderTop: '1px solid transparent',
    borderTopColor: 'inherit',
    transform: 'translateY(50%)',
    boxSizing: 'border-box',
    content: '""',
  },
  '::after': {
    position: 'relative',
    top: '50%',
    width: '50%',
    borderTop: '1px solid transparent',
    borderTopColor: 'inherit',
    transform: 'translateY(50%)',
    boxSizing: 'border-box',
    content: '""',
  },
});

export const sxLeft = stylex.create({
  '::before': {
    top: '50%',
    width: '5%',
  },
  '::after': {
    top: '50%',
    width: '95%',
  },
});

export const sxRight = stylex.create({
  '::before': {
    top: '50%',
    width: '95%',
  },
  '::after': {
    top: '50%',
    width: '5%',
  },
});

const sxHorizontal = stylex.create({
  display: 'flex',
  clear: 'both',
  width: '100%',
  minWidth: '100%',
  margin: '24px 0',
});
const sxVertical = stylex.create({
  position: 'relative',
  top: '-.06em',
  display: 'inline-block',
  height: '.9em',
  margin: '0 8px',
  verticalAlign: 'middle',
  borderTop: 0,
  borderLeft: `1px solid ${$token['line-divider-default']}`,
});

export const sxDirection = (type: 'horizontal' | 'vertical') => {
  if (type === 'horizontal') {
    return sxHorizontal;
  } else {
    if (type === 'vertical') {
      return sxVertical;
    }
  }
};

export const sxDashed = stylex.create({
  borderTopStyle: 'dashed',
});

export const sxWithTextDashed = stylex.create({
  '::before': {
    borderTopStyle: 'dashed',
  },
  '::after': {
    borderTopStyle: 'dashed',
  },
});

export const sxTitle = stylex.create({
  display: 'inline-block',
  padding: '0 1em',
  boxSizing: 'border-box',
});
