// @byted-service/bytedtrace/lib/confProvider/remote/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TCCClient = void 0;
const tslib_1 = require("tslib");
const consts = tslib_1.__importStar(require("../consts"));
const singleflight_1 = tslib_1.__importDefault(require("@byted-service/singleflight"));
const tcc_general_1 = tslib_1.__importStar(require("@byted-service/tcc-general"));
// Client is a wrap of tcc client.
class TCCClient {
    /**
     * 在同一个进程中 psm cluster framework 应该是一致的
     * init default wrapped client with meta info.
     * @param psm
     * @param cluster
     * @param framework
     */
    static init(psm, cluster, framework) {
        this.tccServicePSM = psm;
        this.tccServiceCluster = cluster;
        this.tccServiceFramework =
            typeof framework === 'string' && framework !== ''
                ? framework
                : consts.tccServiceDefaultFramework;
        // 提前 init tcc client
        this.getDefaultClient();
    }
    // getDefaultClient gets the default wrapped client, create new one if client is nil.
    static async getDefaultClient() {
        // Check default client is `nil` or not.
        if (this.defaultClient) {
            return this.defaultClient;
        }
        // 只需要一个 client
        this.defaultClient = await this.sf.do('bytedtrace_new_tcc_client', async () => {
            return await this.newClient(this.tccServicePSM, this.tccServiceCluster, this.tccServiceFramework);
        });
        return this.defaultClient;
    }
    // newClient creates wrapped tcc client by meta info.
    static async newClient(psm, cluster, framework) {
        const tccNodeNameList = [];
        if (typeof psm === 'string' && psm !== '') {
            if (typeof cluster === 'string' && cluster !== '') {
                // TCC group key: /global/go/framework/psm/cluster.
                tccNodeNameList.push(`${consts.tccServiceLanguageNodeName}/${framework}/${psm}/${cluster}`);
            }
            // TCC group key: /global/go/framework/psm.
            tccNodeNameList.push(`${consts.tccServiceLanguageNodeName}/${framework}/${psm}`);
        }
        // TCC group key: /global/go/framework.
        tccNodeNameList.push(`${consts.tccServiceLanguageNodeName}/${framework}`);
        // TCC group key: /global/go.
        tccNodeNameList.push(`${consts.tccServiceLanguageNodeName}`);
        const _tcc = await tcc_general_1.default.create({
            serviceName: consts.tccPlatformServiceName,
            getTimeout: consts.tccDefaultGetTimeout,
            cacheTime: consts.tccDefaultCacheTime,
            disableLog: consts.DISABLE_TCC_GEN_LOG,
        });
        return new TCCClient(tccNodeNameList, _tcc);
    }
    constructor(tccNodeNameList, tccClient) {
        this.tccNodeNameList = tccNodeNameList;
        this.tccClient = tccClient;
    }
    // get gets value by node name & key, will retries all nodes defined in node list until success.
    async getTccConfigByKey(key) {
        for (let i = 0; i < this.tccNodeNameList.length; i++) {
            try {
                const tccNodeName = this.tccNodeNameList[i];
                const value = await this.getWithRetry(tccNodeName, key);
                return value;
            }
            catch (error) /* istanbul ignore next */ {
                if (i === this.tccNodeNameList.length - 1) {
                    throw new Error(error.msg || `tccClient get key error key: ${key}`);
                }
                else {
                    if (error.code !== 100) {
                        throw new Error(error.msg || `tccClient get key error key: ${key}`);
                    }
                }
            }
        }
    }
    // getWithRetry gets tcc config by node name & key with retry.
    async getWithRetry(nodeName, key) {
        for (let i = 0; i < consts.tccDefaultRetryCount; i++) {
            try {
                const res = await this.tccClient.getGroupKey(nodeName, key);
                return res.data;
            }
            catch (error) /* istanbul ignore next */ {
                // 重试结束前只抛出errKeyNotFound错误
                if (i === consts.tccDefaultRetryCount - 1) {
                    throw new tcc_general_1.TCCError(error.code, error.msg);
                }
                else {
                    if (error.code === 100) {
                        // TccErrorCode.errKeyNotFound
                        throw new tcc_general_1.TCCError(error.code, error.msg);
                    }
                }
            }
        }
    }
}
exports.TCCClient = TCCClient;
TCCClient.tccServicePSM = '';
TCCClient.tccServiceCluster = '';
TCCClient.tccServiceFramework = '';
TCCClient.sf = new singleflight_1.default();
