// @byted-service/bytedtrace/lib/confProvider/remote/fetcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Fetcher = void 0;
const tslib_1 = require("tslib");
const gUtils = tslib_1.__importStar(require("../../utils"));
const consts = tslib_1.__importStar(require("../consts"));
const configer = tslib_1.__importStar(require("./configers"));
const promise_allsettled_1 = tslib_1.__importDefault(require("promise.allsettled"));
const lodash_clonedeep_1 = tslib_1.__importDefault(require("lodash.clonedeep"));
class Fetcher {
    constructor(namespace, kvStore, defaultConfig) {
        this.cachedConfig = {};
        this.configer = new configer.CONFIGER_CLASS[namespace](kvStore);
        this.cachedConfig = defaultConfig;
        this.namespace = namespace;
    }
    async fetch() {
        // 如果第一次 result 的 status 不是 fulfilled， 那么会返回默认的数据
        // 如果之前拉取成功，当前拉取失败则返回上一次结果
        const defaultConfig = (0, lodash_clonedeep_1.default)(this.cachedConfig);
        const promises = this.getPromiseByNamespace();
        const results = await promise_allsettled_1.default.call(undefined, promises.map((p) => p.promise));
        results.forEach((result, index) => {
            const key = promises[index].key;
            if (key instanceof Array) {
                if (defaultConfig[key[0]]) {
                    defaultConfig[key[0]][key[1]] = gUtils.ifElse(result.value[1], result.value[0], this.cachedConfig[key[0]][key[1]]);
                }
            }
            else {
                defaultConfig[key] = gUtils.ifElse(result.value[1], result.value[0], this.cachedConfig[key]);
            }
        });
        this.cachedConfig = defaultConfig;
        return defaultConfig;
    }
    getPromiseByNamespace() {
        let promises = [];
        switch (this.namespace) {
            case consts.NAMESPACE_TRACER: {
                const configer = this.configer;
                promises = [
                    {
                        key: 'enableTrace',
                        promise: configer.getGlobalEnableTrace(),
                    },
                    {
                        key: 'enablePostTrace',
                        promise: configer.getGlobalEnablePostTrace(),
                    },
                    {
                        key: 'enableDyedHeader',
                        promise: configer.getGlobalEnableDyedHeader(),
                    },
                    {
                        key: 'enableSampling',
                        promise: configer.getGlobalEnableSampling(),
                    },
                    {
                        key: 'enableReportTotalSpan',
                        promise: configer.getGlobalEnableReportTotalSpan(),
                    },
                    {
                        key: 'spanMetricsLevel',
                        promise: configer.getGlobalSpanMetricsLevel(),
                    },
                    {
                        key: 'customMetricsLevel',
                        promise: configer.getGlobalCustomMetricsLevel(),
                    },
                    {
                        key: 'lightModeByComponent',
                        promise: configer.getGlobalLightModeByComponent(),
                    },
                    {
                        key: 'rootLightModeByComponent',
                        promise: configer.getGlobalRootLightModeByComponent(),
                    },
                    {
                        key: 'disableThriftStyleMetricsByComponent',
                        promise: configer.getGlobalDisableThriftStyleMetricsByComponent(),
                    },
                    {
                        key: 'rootDisableThriftStyleMetricsByComponent',
                        promise: configer.getGlobalRootDisableThriftStyleMetricsByComponent(),
                    },
                ];
                break;
            }
            case consts.NAMESPACE_CB:
                const configer = this.configer;
                promises = [
                    {
                        key: 'maxSpanCountInTxn',
                        promise: configer.getCBMaxSpanCountInTxn(),
                    },
                    {
                        key: 'maxTagKVCountInObj',
                        promise: configer.getCBMaxTagKVCountInObj(),
                    },
                    {
                        key: 'maxEventCountInSpan',
                        promise: configer.getCBMaxEventCountInSpan(),
                    },
                    {
                        key: 'maxCustomMetricCountInSpan',
                        promise: configer.getCBMaxCustomMetricCountInSpan(),
                    },
                    {
                        key: 'maxTagValueLen',
                        promise: configer.getCBMaxTagValueLen(),
                    },
                    {
                        key: 'maxEventDataLen',
                        promise: configer.getCBMaxEventDataLen(),
                    },
                    {
                        key: 'maxTxnDataLen',
                        promise: configer.getCBMaxTxnDataLen(),
                    },
                    {
                        key: 'maxTxnSamplingRatePerOperation',
                        promise: configer.getCBMaxTxnSamplingRatePerOperation(),
                    },
                    {
                        key: 'maxPostTraceSamplingRatePerOperation',
                        promise: configer.getCBMaxPostTraceSamplingRatePerOperation(),
                    },
                    {
                        key: 'maxDyeingSamplingRatePerOperation',
                        promise: configer.getCBMaxDyeingSamplingRatePerOperation(),
                    },
                    {
                        key: 'maxPostTraceSamplingDetailStrategies',
                        promise: configer.getCBMaxPostTraceSamplingDetailStrategies(),
                    },
                ];
                break;
            case consts.NAMESPACE_SAMPLER: {
                const configer = this.configer;
                promises = [
                    {
                        key: ['defaultStrategy', 'samplingMode'],
                        promise: configer.getSamplerSamplingMode(),
                    },
                    {
                        key: ['defaultStrategy', 'samplingRateProbability'],
                        promise: configer.getSamplerSamplingRateProbability(),
                    },
                    {
                        key: ['defaultStrategy', 'samplingCountPerMinAdaptive'],
                        promise: configer.getSamplerSamplingCountPerMinAdaptive(),
                    },
                    {
                        key: 'samplerMaxOperationCount',
                        promise: configer.getSamplerMaxOperationCount(),
                    },
                    {
                        key: 'detailStrategies',
                        promise: configer.getSamplerDetailSamplingStrategies(),
                    },
                    {
                        key: 'conditionSamplingFilters',
                        promise: configer.getSamplerConditionSamplingFilters(),
                    },
                    {
                        key: 'lowerBoundSamplerExpireSecond',
                        promise: configer.getLowerBoundSamplerExpireSecond(),
                    },
                ];
                break;
            }
        }
        return promises;
    }
}
exports.Fetcher = Fetcher;
