// @byted-nodex/metrics-codec/lib/message.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Message = void 0;
const metric_1 = require("./metric");
const util_1 = require("./util");
const message_header_1 = require("./message_header");
const logger_1 = __importDefault(require("./logger"));
const consts_1 = require("./consts");
const typings_1 = require("./typings");
// Message
//+------------+---------+-----+--------+
//| Msg header | Metric  | ... | Metric |
//+------------+---------+-----+--------+
class Message {
    // headerBuffer: Buffer;
    // metricBuffers: Buffer[] = [];
    constructor(options = {}) {
        this.metrics = [];
        this.dc = options.dc || '';
        this.host = options.host || '';
        this.header = {
            magicCode: consts_1.MAGIC_CODE,
            version: consts_1.VERSION,
            bodyOffset: 0,
            bodyLen: 0,
            bodyCompressStrategy: typings_1.CompressStrategy.NO_COMPRESSION,
            bodyCompressLen: 0,
            timestamp: 0n,
            flags: consts_1.DEFAULT_FLAGS,
            metricNum: 0,
            tenant: (0, util_1.checkAndSliceStr)(options.tenant || consts_1.DEFAULT_TENANT),
            account: (0, util_1.checkAndSliceStr)(options.account || consts_1.DEFAULT_ACCOUNT),
            project: (0, util_1.checkAndSliceStr)(options.project || consts_1.DEFAULT_PROJECT),
            sdkVersion: options.sdkVersion || consts_1.DEFAULT_SDK_VERSION,
        };
        // 如果 tenant/account/project/sdkVersion 变了，需要重新计算
        this.header.bodyOffset = (0, message_header_1.getMessageHeaderSize)(this.header);
    }
    get byteLength() {
        return this.header.bodyOffset + this.header.bodyLen;
    }
    addMetric(metricType, metricName, fields, measurements) {
        if (fields.length > consts_1.maxFieldCount) {
            logger_1.default.warn('%s has too many fields: %d, truncated to %d', metricName, fields.length, consts_1.maxFieldCount);
            fields = fields.slice(0, consts_1.maxFieldCount);
        }
        const metric = (0, metric_1.createMetric)(metricType, metricName, fields, measurements);
        const newLength = this.metrics.push(metric);
        this.header.bodyLen += metric.header.metricLen;
        this.header.metricNum = newLength;
        return this;
    }
    encode() {
        if (!this.metrics.length) {
            return;
        }
        let offset = 0;
        const buf = Buffer.allocUnsafe(this.byteLength);
        offset = (0, message_header_1.encodeMessageHeader)(buf, offset, this.header);
        for (const metric of this.metrics) {
            offset = (0, metric_1.encodeMetric)(buf, offset, metric);
        }
        return buf;
    }
    reset() {
        this.metrics.length = 0;
        this.setTimestamp(0n);
        this.setFlags(consts_1.DEFAULT_FLAGS);
        this.header.bodyLen = 0;
        this.header.metricNum = 0;
    }
    // 注意：单位是秒
    setTimestamp(timestamp) {
        this.header.timestamp = timestamp;
    }
    setFlags(flags) {
        this.header.flags = flags;
    }
}
exports.Message = Message;
