// @byted-nodex/metrics-codec/lib/metric.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMetricSize = exports.getTagsSize = exports.encodeTagBytes = exports.encodeTags = exports.encodeMetric = exports.encodeValue = exports.calcMetricHashCode = exports.createMetric = void 0;
const fnv32_1 = require("./fnv32");
const util_1 = require("./util");
const lru_cache_1 = require("lru-cache");
const metricNameCache = new lru_cache_1.LRUCache({ max: 1000 });
const float64bitsCache = new lru_cache_1.LRUCache({ max: 10000 });
function createMetric(metricType, metricName, fields, measurements) {
    const m = {
        header: {
            metricLen: 0,
            num: measurements.length,
            metricType,
            hashcode: calcMetricHashCode(metricName),
            metricName,
            fields,
        },
        measurements,
    };
    m.header.metricLen = getMetricSize(m).size;
    return m;
}
exports.createMetric = createMetric;
function calcMetricHashCode(metricName) {
    if (metricNameCache.has(metricName)) {
        return metricNameCache.get(metricName);
    }
    const hash = (0, fnv32_1.fnv1a)(metricName);
    metricNameCache.set(metricName, hash);
    return hash;
}
exports.calcMetricHashCode = calcMetricHashCode;
function encodeValue(buf, offset, v) {
    let floatValue;
    if (float64bitsCache.has(v)) {
        floatValue = float64bitsCache.get(v);
    }
    else {
        floatValue = (0, util_1.float64bits)(v);
        float64bitsCache.set(v, floatValue);
    }
    return (0, util_1.encodeDouble)(buf, offset, floatValue);
}
exports.encodeValue = encodeValue;
function encodeMetric(buf, offset, metric) {
    // encode metric header
    buf.writeInt32LE(metric.header.metricLen, offset);
    buf.writeInt32LE(metric.measurements.length, offset + 4);
    buf.writeInt8(metric.header.metricType, offset + 8);
    buf.writeUInt32LE(metric.header.hashcode, offset + 9);
    buf.writeInt32LE(metric.header.metricName.length, offset + 13);
    buf.write(metric.header.metricName, offset + 17);
    offset += 17 + metric.header.metricName.length;
    offset = (0, util_1.encodeInt32)(buf, offset, metric.header.fields.length);
    metric.header.fields.forEach(field => {
        offset = (0, util_1.encodeAsciiStr)(buf, offset, field.name);
        offset = (0, util_1.encodeInt8)(buf, offset, field.type);
    });
    // encode measurements
    metric.measurements.forEach(measurement => {
        // tag
        if (measurement.tags instanceof Buffer) {
            measurement.tags.copy(buf, offset);
            offset += measurement.tags.length;
        }
        else {
            offset = encodeTags(buf, offset, measurement.tags);
        }
        // value
        offset = (0, util_1.encodeInt32)(buf, offset, measurement.values.length);
        measurement.values.forEach(value => {
            offset = encodeValue(buf, offset, value);
        });
    });
    return offset;
}
exports.encodeMetric = encodeMetric;
function encodeTags(buf, offset, tags) {
    const hashcode = (0, fnv32_1.fnv32aTags)(tags);
    // encode tags
    offset = (0, util_1.encodeUInt32)(buf, offset, hashcode);
    offset = (0, util_1.encodeInt32)(buf, offset, tags.length * 2);
    tags.forEach(tag => {
        offset = (0, util_1.encodeAsciiStr)(buf, offset, tag.key);
        offset = (0, util_1.encodeAsciiStr)(buf, offset, tag.value);
    });
    return offset;
}
exports.encodeTags = encodeTags;
function encodeTagBytes(tags) {
    const buf = Buffer.allocUnsafe(getTagsSize(tags));
    encodeTags(buf, 0, tags);
    return buf;
}
exports.encodeTagBytes = encodeTagBytes;
function getTagsSize(tags) {
    let size = 8;
    tags.forEach(tag => {
        size += 4 + tag.key.length + 4 + tag.value.length;
    });
    return size;
}
exports.getTagsSize = getTagsSize;
function getMetricSize(metric) {
    const headerSize = 13 + 4 + metric.header.metricName.length + 4;
    let fieldSize = 0;
    metric.header.fields.forEach(field => {
        fieldSize += 4 + field.name.length + 1;
    });
    let measurementSize = 0;
    metric.measurements.forEach(measurement => {
        if (measurement.tags instanceof Buffer) {
            measurementSize += measurement.tags.length;
        }
        else {
            measurementSize += getTagsSize(measurement.tags);
        }
        measurementSize += 4 + measurement.values.length * 12;
    });
    return {
        headerSize,
        fieldSize,
        measurementSize,
        size: headerSize + fieldSize + measurementSize,
    };
}
exports.getMetricSize = getMetricSize;
