// @universe-design/react/lib/pagination/Pagination.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Pagination = void 0;
var _u12o6wmt = _interopRequireDefault(require("./../.stylex/u12o6wmt"));
var _u1rkqst = _interopRequireDefault(require("./../.stylex/u1rkqst1"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _internal = _interopRequireDefault(require("./internal"));
var _clsx = _interopRequireDefault(require("clsx"));
var _LeftBoldOutlined = _interopRequireDefault(require("@universe-design/icons-react/LeftBoldOutlined"));
var _RightBoldOutlined = _interopRequireDefault(require("@universe-design/icons-react/RightBoldOutlined"));
var _UpTopOutlined = _interopRequireDefault(require("@universe-design/icons-react/UpTopOutlined"));
var _MoreOutlined = _interopRequireDefault(require("@universe-design/icons-react/MoreOutlined"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _PaginationSelect = require("./PaginationSelect");
var _useBreakpoint = require("../grid/hooks/useBreakpoint");
var _hooks = require("../config-provider/hooks");
var _util = require("../util");
var _PaginationContext = require("./PaginationContext");
require("./style");
//.u1k6fduh{cursor:pointer;}
//.u1fssmi4{color:@icon-n2;}
/**
 * @docgen
 */
const usePaginationCssScope = () => (0, _hooks.useCssScope)('pagination');
const sxEllipsisIcon = {
  "props": {
    "sj55zd": _u1rkqst.default
  }
};
const sxEllipsis = {
  "props": {
    "Bceei9c": _u12o6wmt.default
  }
};
const Ellipsis = _ref => {
  let {
    position
  } = _ref;
  const prefixCls = usePaginationCssScope()('');
  const {
    ellipsisReadOnly
  } = (0, _PaginationContext.usePaginationContext)();
  if (ellipsisReadOnly) {
    return /*#__PURE__*/_react.default.createElement(_MoreOutlined.default, {
      className: (0, _stylex.default)(sxEllipsisIcon)
    });
  }
  return /*#__PURE__*/_react.default.createElement("a", {
    className: (0, _clsx.default)(`${prefixCls}-item-link`, (0, _stylex.default)(sxEllipsis))
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: `${prefixCls}-item-container`
  }, /*#__PURE__*/_react.default.createElement(_UpTopOutlined.default
  // FIXME 修改为正常的 icon
  , {
    style: {
      transform: position === 'start' ? 'rotate(-90deg)' : 'rotate(90deg)'
    },
    className: `${prefixCls}-item-link-icon`
  }), /*#__PURE__*/_react.default.createElement("span", {
    className: `${prefixCls}-item-ellipsis`
  }, /*#__PURE__*/_react.default.createElement(_MoreOutlined.default, null))));
};

/**
 * @public
 */
const Pagination = _ref2 => {
  let {
    className,
    size,
    itemRender,
    simple,
    showQuickJumper,
    showSizeChanger,
    showLessItems,
    ...restProps
  } = _ref2;
  process.env.NODE_ENV !== "production" ? (0, _util.warning)(typeof simple !== 'boolean', (0, _util.compatibleDevWarning)('Pagination', ['simple'])) : void 0;
  process.env.NODE_ENV !== "production" ? (0, _util.warning)(typeof showLessItems === 'undefined', (0, _util.deprecatedWarning)('Pagination', 'showLessItems', ['siblingCount', 'boundaryCount'])) : void 0;
  const {
    xs
  } = (0, _useBreakpoint.useBreakpoint)();
  const direction = (0, _hooks.useDirection)();
  const prefixCls = (0, _hooks.useCssScope)('pagination')('');
  const selectPrefixCls = (0, _hooks.useCssScope)('select')('');
  const locale = (0, _hooks.useLocaleReceiver)('pagination');
  const isSmall = size === 'small' || Boolean(xs && !size && restProps.responsive);
  const extendedClassName = (0, _clsx.default)({
    mini: isSmall,
    [`${prefixCls}-rtl`]: direction === 'rtl'
  }, className);
  const mergedShowQuickJumper = showQuickJumper !== null && showQuickJumper !== void 0 ? showQuickJumper : false;
  const mergedShowSizeChanger = // eslint-disable-next-line @typescript-eslint/no-magic-numbers
  showSizeChanger !== null && showSizeChanger !== void 0 ? showSizeChanger : Boolean(!isSmall && typeof restProps.total === 'number' && restProps.total > 50);
  return /*#__PURE__*/_react.default.createElement(_internal.default, (0, _extends2.default)({}, restProps, {
    prefixCls: prefixCls,
    selectPrefixCls: selectPrefixCls
  }, (() => {
    let prevIcon = /*#__PURE__*/_react.default.createElement("button", {
      className: `${prefixCls}-item-link`,
      type: "button",
      tabIndex: -1
    }, /*#__PURE__*/_react.default.createElement(_LeftBoldOutlined.default, null));
    let nextIcon = /*#__PURE__*/_react.default.createElement("button", {
      className: `${prefixCls}-item-link`,
      type: "button",
      tabIndex: -1
    }, /*#__PURE__*/_react.default.createElement(_RightBoldOutlined.default, null));
    let jumpPrevIcon = /*#__PURE__*/_react.default.createElement(Ellipsis, {
      position: "start"
    });
    let jumpNextIcon = /*#__PURE__*/_react.default.createElement(Ellipsis, {
      position: "end"
    });
    // change arrows direction in right-to-left direction
    if (direction === 'rtl') {
      [prevIcon, nextIcon] = [nextIcon, prevIcon];
      [jumpPrevIcon, jumpNextIcon] = [jumpNextIcon, jumpPrevIcon];
    }
    return {
      prevIcon,
      nextIcon,
      jumpPrevIcon,
      jumpNextIcon
    };
  })(), {
    className: extendedClassName,
    selectComponentClass: _PaginationSelect.PaginationSelect,
    locale: locale,
    itemRender: itemRender,
    showQuickJumper: mergedShowQuickJumper,
    showSizeChanger: mergedShowSizeChanger
  }));
};
exports.Pagination = Pagination;
if (process.env.NODE_ENV !== "production") {
  Pagination.displayName = 'Pagination';
}
