// @universe-design/react/src/switch/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Basic from '../demo/0.basic';
import C1Disabled from '../demo/1.disabled';
import C2Size from '../demo/2.size';
import C3Loading from '../demo/3.loading';
if (!(C0Basic as Story).storyName) {
  (C0Basic as Story).storyName = '0. 基础用法';
  (C0Basic as Story).parameters = {
    docs: {
      description: {
        story: '简单的用法。\n\n',
      },
    },
  };
}
if (!(C1Disabled as Story).storyName) {
  (C1Disabled as Story).storyName = 'null. 不可用';
  (C1Disabled as Story).parameters = {
    docs: {
      description: {
        story: '开关失效状态。\n\n',
      },
    },
  };
}
if (!(C2Size as Story).storyName) {
  (C2Size as Story).storyName = '2. 尺寸';
  (C2Size as Story).parameters = {
    docs: {
      description: {
        story: '两种不同尺寸的开关。\n\n',
      },
    },
  };
}
if (!(C3Loading as Story).storyName) {
  (C3Loading as Story).storyName = '3. 加载中';
  (C3Loading as Story).parameters = {
    docs: {
      description: {
        story: '标识开关操作仍在执行中。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Switch/官网示例',
  tags: ['autodocs'],
};

export { C0Basic, C1Disabled, C2Size, C3Loading };
