// @universe-design/react/src/expandable-text/Expandable.tsx
import * as React from 'react';
import { useImmer } from 'use-immer';
import { useEventCallback } from '../hooks';
import { createContextHook, useControlledState } from '@universe-design/react-utilities';

interface ControllableProp<T> {
  /**
   * 受控值
   */
  value?: T;
  /**
   * 默认值
   */
  defaultValue?: T;
  /**
   * 值变化回调
   */
  onChange?: (value: T) => unknown;
}

export interface ExpandableProps {
  /**
   * 是否展开
   * @defaultValue false
   */
  isExpanded?: ControllableProp<boolean>;
  /**
   * 是否在展开后阻止收起
   * @defaultValue false
   */
  preventCollapseWhenExpanded?: boolean;
}

const useExpandable = (props: ExpandableProps) => {
  const { isExpanded: isExpandedProp = {}, preventCollapseWhenExpanded = false } = props;

  /**
   * State
   */
  const [{ isOverflowed }, updateState] = useImmer({
    isOverflowed: false,
  });

  /**
   * isExpanded State
   */
  const [isExpanded, setExpanded] = useControlledState(isExpandedProp.defaultValue ?? false, isExpandedProp.value, {
    onChange: isExpandedProp.onChange,
  });

  const toggleExpanded = useEventCallback(() => {
    setExpanded(!isExpanded);
  });

  const updateOverflowed = useEventCallback((_isOverflowed: boolean) => {
    updateState(draft => {
      draft.isOverflowed = _isOverflowed;
    });
  });

  const shouldHideTrigger = (isExpanded && preventCollapseWhenExpanded) || !isOverflowed;

  return {
    /**
     * 是否隐藏 trigger
     */
    shouldHideTrigger,
    isExpanded,
    toggleExpanded,
    updateOverflowed,
  };
};

const { Provider: ExpandableContextProvider, useContext: useExpandableContext } = createContextHook<
  ReturnType<typeof useExpandable>
>('ExpandableContext', {
  strict: true,
});

export const ExpandableProvider: React.FC<ExpandableProps> = ({ children, ...props }) => (
  <ExpandableContextProvider value={useExpandable(props)}>{children}</ExpandableContextProvider>
);

export { useExpandableContext };
