// @universe-design/react/src/expandable-text/ExpandableText.tsx
/** @jsxRuntime classic */
/** @jsx jsx */
import stylex from '@universe-design/stylex';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx, useMergeComponentProps } from '@universe-design/stylex/react';
import * as React from 'react';
import clamp from 'lodash/clamp';
import DownOutlined from '@universe-design/icons-react/DownOutlined';
import { useImmer } from 'use-immer';
import { useCssScope, useDirection, useLocaleReceiver } from '../config-provider';
import type { StylexProps } from '../util';
import { forwardHTMLRef } from '../util/forwardRef';
import { Button } from '../button';
import { EmptyElement } from '../shared/components/FirstLineAlign';
import { useIsomorphicLayoutEffect as useLayoutEffect } from '../hooks';
import { useResizeObserver } from '@universe-design/react-utilities';
import type { ExpandableProps } from './Expandable';
import { ExpandableProvider, useExpandableContext } from './Expandable';

import './style/text-collapse.less';

const useExpandableTextCssScope = () => useCssScope('expandable-text');

export interface ExpandableTextParagraphExoticProps {
  /**
   * 收起时最多显示几行文本
   */
  lineClamp?: number;
}

/**
 * @public
 * @docgen
 */
export interface ExpandableTextTriggerExoticProps {
  // /**
  //  * 是否为加载态。
  //  * @defaultValue false
  //  */
  // loading?: boolean;
  /**
   * 蒙层的背景颜色。用于给触发器设置与文字区域同样的背景颜色，只在触发器展示为悬停元素的时候有效。
   * @defaultValue `var(--bg-body)`
   */
  maskColor?: string;
  /**
   * 触发器的位置。
   * - `start`: 出现在行首，始终为行级元素。
   * - `end`: 出现在行尾，收起时为悬停元素，展开时为行级元素。
   * @defaultValue `end`
   */
  position?: 'start' | 'end';
}

export interface ExpandableTextExoticProps
  extends StylexProps,
    Pick<ExpandableProps, 'preventCollapseWhenExpanded'>,
    ExpandableTextParagraphExoticProps {
  /**
   * 触发器元素的属性
   */
  triggerProps?: ExpandableTextTriggerExoticProps;
}

const ExpandableTextParagraph: React.FC<ExpandableTextParagraphExoticProps> = props => {
  const { lineClamp: lineClampProp = 3, children } = props;
  const lineClamp = clamp(lineClampProp, 1, Infinity);
  const { isExpanded, updateOverflowed } = useExpandableContext();
  const cssScope = useExpandableTextCssScope();
  const paragraphRef = useResizeObserver<HTMLDivElement>(() => {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    reCalcLineBottoms();
  });
  const paragraphContentRef = React.useRef<HTMLSpanElement>(null);

  /**
   * State
   */
  const [{ lineBottoms }, updateState] = useImmer({
    lineBottoms: [0] as number[],
  });

  const reCalcLineBottoms = React.useCallback(() => {
    const paragraph = paragraphRef.current;
    const paragraphContent = paragraphContentRef.current;
    if (!paragraph || !paragraphContent) {
      return;
    }
    // 获取真实占位的 boundary
    const { top, bottom, height } = paragraph.getBoundingClientRect();
    const scaleY = height / paragraph.offsetHeight;
    const lineRects = [...paragraphContent.getClientRects()].filter(
      rect => rect.height && rect.width, // firefox 在子元素是 div 时会多一个 width 是 0 的 rect，我们要排除掉
    );
    // console.trace('reCalcLineBottoms', { scaleY, top, bottom, lineRects });
    updateState(draft => {
      let prevLine = {
        index: 0,
        bottom: 0,
      };
      draft.lineBottoms.splice(1 + lineRects.length);
      draft.lineBottoms[0] = 0;
      for (const [i, rect] of lineRects.entries()) {
        const spacingBottom = (() => {
          // eslint-disable-next-line max-nested-callbacks
          const nextLine = lineRects.slice(i + 1).find(r => r.top > rect.top); // 获取当前 rect 后第一个不同行的 rect，如果找到则把两行间隔的一半作为分界
          if (nextLine) {
            return (nextLine.top - rect.bottom) / 2;
          }
          return bottom - rect.bottom; // 如果没有找到不同行的 rect 说明是最后一行了，这时用 wrapper 的 bottom 作为边界
        })();
        /**
         * currentLineBottom 是相对于 top 的底部坐标，相当于 height，可以直接除 scaleY 得到 layout 尺寸
         */
        const currentLineBottom = Math.round((rect.bottom - top + spacingBottom) / scaleY); // 计算出当前行的下边界
        if (currentLineBottom === prevLine.bottom) {
          continue; // 如果和上次计算出的行下边界一致则说明这次的 rect 与之前的 rect 同行，这种情况出现在 white-space: pre-line 时，换行符 \n 会产生一个同行的零宽 rect
        }
        prevLine = {
          index: prevLine.index + 1,
          bottom: currentLineBottom,
        };
        draft.lineBottoms[prevLine.index] = prevLine.bottom;
        // console.log('draft.lineBottoms', JSON.stringify(draft.lineBottoms));
      }
    });
    updateOverflowed(lineRects.length > lineClamp);
  }, [lineClamp, paragraphRef, updateOverflowed, updateState]);

  // lineClamp 变化的时候可以保证 reactive
  useLayoutEffect(() => {
    reCalcLineBottoms();
  }, [reCalcLineBottoms]);

  return (
    <div
      className={cssScope('__paragraph')}
      sx={[
        stylex.create({ margin: 0 }),
        !isExpanded &&
          stylex.create({
            overflowY: 'hidden',
          }),
      ]}
      style={{
        ...(isExpanded
          ? {}
          : {
              maxHeight: lineBottoms[lineClamp] ?? lineBottoms[lineBottoms.length - 1], // 这里不直接用 webkit-line-clamp 实现是因为 safari 最后一行的底部没有 spacing 导致 trigger 定位会偏上
            }),
      }}
    >
      <div ref={paragraphRef}>
        <span ref={paragraphContentRef} className={cssScope('__paragraph__content')}>
          {children}
        </span>
      </div>
    </div>
  );
};

const ExpandableTextTrigger: React.FC<ExpandableTextTriggerExoticProps> = props => {
  const { maskColor, position: triggerPosition = 'end' } = props;
  const cssScope = useExpandableTextCssScope();
  const direction = useDirection();
  const { isExpanded, shouldHideTrigger, toggleExpanded } = useExpandableContext();
  const locale = useLocaleReceiver('expandableText');

  const shouldShowMask = triggerPosition === 'end' && !isExpanded;

  /**
   * Render
   */
  const maskNode = React.useMemo(() => {
    if (shouldShowMask) {
      return (
        <div
          className={cssScope('__trigger__mask')}
          sx={[
            stylex.create({
              display: 'flex',
              width: 16,
              '-webkit-mask-image': 'linear-gradient(90deg, rgba(0, 0, 0, 0) 0%, rgba(0, 0, 0, 1) 100%)',
              backgroundColor: $token['bg-body'],
            }),
          ]}
          style={{ backgroundColor: maskColor }}
        >
          <EmptyElement />
        </div>
      );
    }
    return null;
  }, [maskColor, cssScope, shouldShowMask]);

  const buttonNode = React.useMemo(
    () => (
      <div
        sx={[
          stylex.create({
            display: 'flex',
          }),
          shouldShowMask &&
            stylex.create({
              backgroundColor: $token['bg-body'],
            }),
        ]}
        style={{
          ...(shouldShowMask && { backgroundColor: maskColor }),
        }}
      >
        <Button
          className={cssScope('__trigger__button')}
          type="link"
          // 这里用 `DownOutlined` 而不是 `DownBoldOutlined` 是因为这里细一点好看
          endIcon={
            <DownOutlined
              style={{
                ...(isExpanded && { transform: 'rotate(180deg)' }),
              }}
            />
          }
          onClick={toggleExpanded}
        >
          {isExpanded ? locale.fold : locale.show}
        </Button>
      </div>
    ),
    [shouldShowMask, maskColor, cssScope, isExpanded, toggleExpanded, locale.fold, locale.show],
  );

  if (shouldHideTrigger) {
    return null;
  }

  return (
    <div
      className={cssScope('__trigger')}
      sx={[
        stylex.create({
          display: 'flex',
        }),
        triggerPosition === 'start' && [
          stylex.create({
            justifyContent: 'flex-start',
          }),
        ],
        triggerPosition === 'end' && [
          isExpanded && [
            stylex.create({
              justifyContent: 'flex-end',
            }),
          ],
          !isExpanded && [
            stylex.create({
              position: 'absolute',
              bottom: 0,
            }),
            direction === 'ltr' ? stylex.create({ right: 0 }) : stylex.create({ left: 0 }),
          ],
        ],
      ]}
    >
      {maskNode}
      {buttonNode}
    </div>
  );
};

/**
 * 文字收起组件
 */
export const ExpandableText = forwardHTMLRef<ExpandableTextExoticProps, 'div'>((props, ref) => {
  const { children, preventCollapseWhenExpanded, lineClamp, triggerProps, ...domProps } = props;

  /**
   * Context
   */
  const cssScope = useExpandableTextCssScope();

  /**
   * Props
   */
  const mergedDOMProps = useMergeComponentProps(domProps, {
    sx: [
      stylex.create(
        {
          position: 'relative', // 展开按钮会用 absolute 布局
        },
        cssScope(),
      ),
    ],
  });

  return (
    <ExpandableProvider {...{ preventCollapseWhenExpanded }}>
      <div ref={ref} {...mergedDOMProps}>
        <ExpandableTextParagraph lineClamp={lineClamp}>{children}</ExpandableTextParagraph>
        <ExpandableTextTrigger {...triggerProps} />
      </div>
    </ExpandableProvider>
  );
});

if (__DEV__) {
  ExpandableText.displayName = 'ExpandableText';
}

/**
 * @public
 * @docgen
 */
export type ExpandableTextProps = React.ComponentProps<typeof ExpandableText>;
