// @universe-design/react/src/loading/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Basic from '../demo/0.basic';
import C1Size from '../demo/1.size';
import C2Tip from '../demo/2.tip';
import C3DelayAndDebounce from '../demo/3.delay-and-debounce';
import C4CustomizedIndicator from '../demo/4.customized-indicator';
import C5RenderPropPattern from '../demo/5.render-prop-pattern';
import C6LoadingIllustration from '../demo/6.loading-illustration';
if (!(C0Basic as Story).storyName) {
  (C0Basic as Story).storyName = '0. 基础用法';
  (C0Basic as Story).parameters = {
    docs: {
      description: {
        story: '简单的加载状态。\n\n',
      },
    },
  };
}
if (!(C1Size as Story).storyName) {
  (C1Size as Story).storyName = '1. 各种大小';
  (C1Size as Story).parameters = {
    docs: {
      description: {
        story: '小的用于文本加载，默认用于卡片容器级加载，大的用于页面级加载。\n\n',
      },
    },
  };
}
if (!(C2Tip as Story).storyName) {
  (C2Tip as Story).storyName = '2. 自定义描述文案';
  (C2Tip as Story).parameters = {
    docs: {
      description: {
        story: '自定义描述文案。\n\n',
      },
    },
  };
}
if (!(C3DelayAndDebounce as Story).storyName) {
  (C3DelayAndDebounce as Story).storyName = '3. 延迟';
  (C3DelayAndDebounce as Story).parameters = {
    docs: {
      description: {
        story: '延迟显示加载效果。当加载状态在 `delay` 时间内结束，则不显示加载状态。\n\n',
      },
    },
  };
}
if (!(C4CustomizedIndicator as Story).storyName) {
  (C4CustomizedIndicator as Story).storyName = '4. 自定义指示符';
  (C4CustomizedIndicator as Story).parameters = {
    docs: {
      description: {
        story: '使用自定义指示符。\n\n',
      },
    },
  };
}
if (!(C5RenderPropPattern as Story).storyName) {
  (C5RenderPropPattern as Story).storyName = '5. 子元素获取加载状态';
  (C5RenderPropPattern as Story).parameters = {
    docs: {
      description: {
        story: '由于加载组件可能存在延迟逻辑，如果子元素需要获取加载状态可以通过 children render prop 模式获取。\n\n',
      },
    },
  };
}
if (!(C6LoadingIllustration as Story).storyName) {
  (C6LoadingIllustration as Story).storyName = '6. Loading 插画';
  (C6LoadingIllustration as Story).parameters = {
    docs: {
      description: {
        story:
          '可以通过 插画 Illustration 组件实现 loading 动画， [点击查看更多插画用法](/develop/react/illustration)。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Loading/官网示例',
  tags: ['autodocs'],
};

export {
  C0Basic,
  C1Size,
  C2Tip,
  C3DelayAndDebounce,
  C4CustomizedIndicator,
  C5RenderPropPattern,
  C6LoadingIllustration,
};
