// @universe-design/react/src/color-picker/pickers/ColorPicker.tsx
import type { FC } from 'react';
import React, { useState, useMemo, useRef } from 'react';
import clsx from 'clsx';
import type { ColorParam, ColorPickerBlocksProps } from '../type';
import { ColorPickerStep } from '../type';
import { judgeEmptyColorGroup } from '../util';
import RightBoldOutlined from '@universe-design/icons-react/RightBoldOutlined';
import { useLocale } from '../../config-provider/hooks';
import { DEFAULT_COLOR_GROUP } from '../colors';
import { Palette } from './Palette';
import { MOREBTNICON } from '../const';
import { useComplementEvent } from '../../hooks/useComplementEvent';
import { ColorPickerProvider, useColorPicker, useColorPickerContext } from '../ColorPickerContext';
import { ColorPickerColorBlock } from '../components/ColorPickerColorBlock';
import type { ColorPickerSaturationProps } from '../components/ColorPickerSaturation';
import type { ColorPickerHueProps } from '../components/ColorPickerHue';

import '../style';

/**
 * @public
 * @docgen
 */
export interface ColorPickerProps
  extends ColorPickerBlocksProps,
    Pick<ColorPickerSaturationProps, 'onSaturationDragEnd'>,
    Pick<ColorPickerHueProps, 'onHueDragEnd'> {
  /**
   * 色块面板的色值选项，可由此配置行列数
   * @defaultValue DEFAULT_COLOR_GROUP
   */
  colorGroup?: ColorParam[][];
  /**
   * 是否显示更多颜色入口
   */
  showMoreColor?: boolean;
  /**
   * 更多颜色入口的文案
   */
  moreBtnText?: string;
}

const Blocks = (props: { colorGroup: ColorParam[][] }) => {
  const { colorGroup } = props;
  const { rootCssScope } = useColorPickerContext();
  const prefix = rootCssScope('-swatches');

  const isEmptyColorGroup = useMemo(() => judgeEmptyColorGroup(colorGroup), [colorGroup]);

  return (
    <div className={clsx(prefix, isEmptyColorGroup && `${prefix}-empty`)}>
      {colorGroup.map((colors, index) => (
        <div key={index} className={`${prefix}-row`}>
          {colors.map(color => (
            <ColorPickerColorBlock key={JSON.stringify(color)} color={color} />
          ))}
        </div>
      ))}
    </div>
  );
};

export const ColorPicker: FC<ColorPickerProps> = props => {
  const locale = useLocale();
  const {
    colorGroup = DEFAULT_COLOR_GROUP,
    showMoreColor = true,
    moreBtnText = locale.components.colorPicker.moreColor,
    className,
    style,
    onSaturationDragEnd,
    onHueDragEnd,
  } = props;

  const [currentStep, setCurrentStep] = useState(ColorPickerStep.inHome);
  const colorPickerCtx = useColorPicker(props);
  const { rootCssScope } = colorPickerCtx;
  const prefix = rootCssScope();

  const isEmptyColorGroup = useMemo(() => judgeEmptyColorGroup(colorGroup), [colorGroup]);

  const containerRef = useRef<HTMLDivElement>(null);
  const moreColorBtnRef = useRef<HTMLDivElement>(null);
  useComplementEvent('pointerdown', {
    listener: React.useCallback(ev => {
      setCurrentStep(ColorPickerStep.inHome);
    }, []),
    // 这里必须有两个 ref，是因为点击 moreColorBtn 的时候会跳转到另一页，在事件触发的过程中，moreColorBtn 已经消失了，所以会判断为点击了外面，这个写法是为了保证事件从 moreColor 出来也可以判断是在组件内部
    elements: React.useMemo(() => [containerRef, moreColorBtnRef.current], []),
    skipped: currentStep !== ColorPickerStep.inMoreColor,
  });

  const renderMoreColorBtn = () => {
    if (showMoreColor === false) {
      return null;
    }

    return (
      <div
        ref={moreColorBtnRef}
        className={clsx(`${prefix}-more`, isEmptyColorGroup && `${prefix}-more-empty`)}
        onClick={() => setCurrentStep(ColorPickerStep.inMoreColor)}
      >
        <div className={`${prefix}-more-left`}>
          <img src={MOREBTNICON} />
          <span>{moreBtnText}</span>
        </div>
        <RightBoldOutlined />
      </div>
    );
  };

  const renderColorPicker = () => {
    if (currentStep === ColorPickerStep.inHome) {
      return (
        <div ref={containerRef} className={clsx(prefix, className)} style={style}>
          <Blocks colorGroup={colorGroup} />
          {renderMoreColorBtn()}
        </div>
      );
    }
    return (
      <Palette
        onSaturationDragEnd={onSaturationDragEnd}
        onHueDragEnd={onHueDragEnd}
        ref={containerRef}
        className={className}
        style={style}
      />
    );
  };

  return <ColorPickerProvider value={colorPickerCtx}>{renderColorPicker()}</ColorPickerProvider>;
};
