// @universe-design/react/src/date-picker/weekNumberWrapper/index.tsx
import React from 'react';
import clsx from 'clsx';
import type moment from 'moment';
import { Tooltip } from '../../tooltip';
import { useCssScope, useLocale } from '../../config-provider/hooks';
import './index.less';

export type WeekNumberWrapperProps = React.ComponentPropsWithoutRef<'div'> & {
  rowNum: number;
  colNum: number;
  baseDate: moment.Moment;
  hideWeekNumbers: boolean;
};

const WeekNumberWrapper = React.forwardRef<HTMLDivElement, WeekNumberWrapperProps>(
  ({ hideWeekNumbers, rowNum, colNum, baseDate, children, ...restProps }, ref) => {
    const cssScope = useCssScope('picker__cell-week-number');
    const wrapperCssScope = useCssScope('picker__wrapper-week-number');
    const headerCssScope = useCssScope('picker__header-week-number');
    const {
      components: { datePicker: locale },
    } = useLocale();
    if (hideWeekNumbers) {
      return <>{children}</>;
    }

    const getWeekNumberRows = () =>
      [...new Array(rowNum)].map((_, rowIndex) => {
        const offset = (rowIndex + 1) * colNum - 1;
        const weekEndDate = baseDate.clone().add(offset, 'day');
        const weekNumber = weekEndDate.weeks();
        const tooltipTitle = locale.weekNumber.f({ num: String(weekNumber) });
        return (
          <div className={clsx(cssScope())} key={rowIndex}>
            <Tooltip title={tooltipTitle}>
              <div className={clsx(cssScope('__inner'))}>{weekNumber}</div>
            </Tooltip>
          </div>
        );
      });

    return (
      <div ref={ref} className={clsx(wrapperCssScope())} {...restProps}>
        <div className={clsx(headerCssScope())}>
          <div className={clsx(headerCssScope('-name'))} />
          <div className={clsx(headerCssScope('-content'))}>{getWeekNumberRows()}</div>
        </div>
        <div className={clsx(headerCssScope('__space'))} />
        {children}
      </div>
    );
  },
);

export default WeekNumberWrapper;
