// @universe-design/react/src/color-picker/stories/homeless.stories.tsx
import React, { useCallback, useState } from 'react';
import type { Meta } from '@storybook/react';
import type { ColorPickerProps } from '..';
import {
  ColorPicker,
  PaletteColorPicker,
  SimpleBlocksColorPicker,
  PrimitiveColorPicker,
  ColorPickerChangeSource,
} from '..';
import { Button, Dropdown } from '@universe-design/react';
import { PresetThemeLight } from '@universe-design/styles';
import { Space } from '../..';

export default {
  title: 'Components/ColorPicker/临时示例',
  component: ColorPicker,
  subcomponents: {
    PaletteColorPicker,
    SimpleBlocksColorPicker,
    PrimitiveColorPicker,
  },
} satisfies Meta;

export const EmptyDemo = () => <ColorPicker colorGroup={[]} />;

export const CloseAfterPick = () => {
  const [visible, setVisible] = React.useState(false);
  const [color, setColor] = React.useState<string | undefined>(PresetThemeLight.B500);
  const onChange: ColorPickerProps['onChange'] = useCallback((c, h, event) => {
    if (event.source === ColorPickerChangeSource.COLOR_BLOCK) {
      setVisible(false);
    }
    setColor(c);
  }, []);

  const close = useCallback(() => {
    setVisible(false);
  }, [setVisible]);

  return (
    <Dropdown
      visible={visible}
      onVisibleChange={setVisible}
      overlay={<ColorPicker onChange={onChange} onHueDragEnd={close} onSaturationDragEnd={close} value={color} />}
      trigger="click"
    >
      <Button onClick={() => setVisible(true)}>选择颜色</Button>
    </Dropdown>
  );
};

export const UndefinedValue = () => {
  const colors = [
    ['#bacefd', '#4e83fd', '#245bdb'],
    ['#faf1d1', '#fad355', '#dc9b04'],
    ['#fbbfbc', '#f76964', '#d83931'],
    ['#000000'],
  ];
  const [value, setValue] = useState<string | undefined>('#fbbfbc');
  return (
    <>
      <Space direction="vertical">
        <ColorPicker colorGroup={colors} defaultValue={undefined} />
        <ColorPicker colorGroup={colors} value={value} onChange={v => setValue(v)} />
        <span>当前值为：{value || 'undefined'}</span>
        <Button onClick={() => setValue(undefined)}>设为 undefined</Button>
        <Button onClick={() => setValue('#000000')}>设为 #000000</Button>
        <Button onClick={() => setValue('#bacefd')}>设为 #bacefd</Button>
        <Button onClick={() => setValue('#4e83fd')}>设为 #4e83fd</Button>
        <Button onClick={() => setValue('#245bdb')}>设为 #245bdb</Button>
      </Space>
    </>
  );
};
