// @universe-design/react/src/color-picker/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Basic from '../demo/0.basic';
import C1Control from '../demo/1.control';
import C2HiddenMore from '../demo/2.hidden-more';
import C3MoreSetting from '../demo/3.more-setting';
import C4Sketch from '../demo/4.sketch';
import C5Dropdown from '../demo/5.dropdown';
import C6PrimitivePicker from '../demo/6.primitive-picker';
if (!(C0Basic as Story).storyName) {
  (C0Basic as Story).storyName = '0. 基础用法';
  (C0Basic as Story).parameters = {
    docs: {
      description: {
        story: '简单的用法。\n\n',
      },
    },
  };
}
if (!(C1Control as Story).storyName) {
  (C1Control as Story).storyName = '1. 受控模式';
  (C1Control as Story).parameters = {
    docs: {
      description: {
        story: '提供`value`让颜色选择器进入受控模式，可配合表单使用。\n\n',
      },
    },
  };
}
if (!(C2HiddenMore as Story).storyName) {
  (C2HiddenMore as Story).storyName = '2. 简单的色块选择器';
  (C2HiddenMore as Story).parameters = {
    docs: {
      description: {
        story: '使用 `SimpleBlocksColorPicker` 组件，实现简单的色块选择器。\n\n',
      },
    },
  };
}
if (!(C3MoreSetting as Story).storyName) {
  (C3MoreSetting as Story).storyName = '3. 自定义颜色选择器';
  (C3MoreSetting as Story).parameters = {
    docs: {
      description: {
        story: '通过`colorGroup`自定义色板，通过`moreBtnText`自定义更多颜色入口的文案。\n\n',
      },
    },
  };
}
if (!(C4Sketch as Story).storyName) {
  (C4Sketch as Story).storyName = '4. 使用调色板';
  (C4Sketch as Story).parameters = {
    docs: {
      description: {
        story: '单独使用调色板。\n\n',
      },
    },
  };
}
if (!(C5Dropdown as Story).storyName) {
  (C5Dropdown as Story).storyName = '4. 下拉选择颜色';
  (C5Dropdown as Story).parameters = {
    docs: {
      description: {
        story: '配合下拉菜单使用。\n\n',
      },
    },
  };
}
if (!(C6PrimitivePicker as Story).storyName) {
  (C6PrimitivePicker as Story).storyName = '5. 自由组合子组件';
  (C6PrimitivePicker as Story).parameters = {
    docs: {
      description: {
        story:
          '使用 `PrimitiveColorPicker` 配合 `ColorPickerColorBlock`, `ColorPickerHue`, `ColorPickerSaturation` 等子组件可以实现更灵活的布局\n\n',
      },
    },
  };
}

export default {
  title: 'Components/ColorPicker/官网示例',
  tags: ['autodocs'],
};

export { C0Basic, C1Control, C2HiddenMore, C3MoreSetting, C4Sketch, C5Dropdown, C6PrimitivePicker };
