// @universe-design/styles/cjs/base/index.dark.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.basetoken_form_moon_03 = exports.basetoken_form_moon_02 = exports.basetoken_form_moon_01 = exports.basetoken_form_moon = exports.basetoken_form_input_bg = exports.basetoken_form_hourglass_04 = exports.basetoken_form_hourglass_03 = exports.basetoken_form_hourglass_02 = exports.basetoken_form_hourglass_01 = exports.basetoken_form_hourglass = exports.basetoken_form_header_bg_press = exports.basetoken_form_header_bg_hover = exports.basetoken_form_header_bg = exports.basetoken_form_goal_04 = exports.basetoken_form_goal_03 = exports.basetoken_form_goal_02 = exports.basetoken_form_goal_01 = exports.basetoken_form_goal = exports.basetoken_form_fog_04 = exports.basetoken_form_fog_03 = exports.basetoken_form_fog_02 = exports.basetoken_form_fog_01 = exports.basetoken_form_fog = exports.basetoken_form_flower_04 = exports.basetoken_form_flower_03 = exports.basetoken_form_flower_02 = exports.basetoken_form_flower_01 = exports.basetoken_form_flower = exports.basetoken_form_entertainment_04 = exports.basetoken_form_entertainment_03 = exports.basetoken_form_entertainment_02 = exports.basetoken_form_entertainment_01 = exports.basetoken_form_entertainment = exports.basetoken_form_desert_04 = exports.basetoken_form_desert_03 = exports.basetoken_form_desert_02 = exports.basetoken_form_desert_01 = exports.basetoken_form_desert = exports.basetoken_form_award_04 = exports.basetoken_form_award_03 = exports.basetoken_form_award_02 = exports.basetoken_form_award_01 = exports.basetoken_form_award = exports.basetoken_dashboard_slicer_tag_bg_normal = exports.basetoken_dashboard_slicer_tag_bg_hover = exports.basetoken_dashboard_sidebar_bg_overlay = exports.basetoken_container_bg = exports.basetoken_AI_instruction_press = exports.basetoken_AI_instruction_normal = exports.basetoken_AI_instruction_hover = void 0;
exports.basetoken_table_bg_turquoise_01 = exports.basetoken_table_bg_red_04 = exports.basetoken_table_bg_red_03 = exports.basetoken_table_bg_red_02 = exports.basetoken_table_bg_red_01 = exports.basetoken_table_bg_purple_04 = exports.basetoken_table_bg_purple_03 = exports.basetoken_table_bg_purple_02 = exports.basetoken_table_bg_purple_01 = exports.basetoken_table_bg_orange_04 = exports.basetoken_table_bg_orange_03 = exports.basetoken_table_bg_orange_02 = exports.basetoken_table_bg_orange_01 = exports.basetoken_table_bg_neutral_04 = exports.basetoken_table_bg_neutral_03 = exports.basetoken_table_bg_neutral_02 = exports.basetoken_table_bg_neutral_01 = exports.basetoken_table_bg_lime_04 = exports.basetoken_table_bg_lime_03 = exports.basetoken_table_bg_lime_02 = exports.basetoken_table_bg_lime_01 = exports.basetoken_table_bg_green_04 = exports.basetoken_table_bg_green_03 = exports.basetoken_table_bg_green_02 = exports.basetoken_table_bg_green_01 = exports.basetoken_table_bg_carmine_04 = exports.basetoken_table_bg_carmine_03 = exports.basetoken_table_bg_carmine_02 = exports.basetoken_table_bg_carmine_01 = exports.basetoken_table_bg_blue_04 = exports.basetoken_table_bg_blue_03 = exports.basetoken_table_bg_blue_02 = exports.basetoken_table_bg_blue_01 = exports.basetoken_home_create_new_bg = exports.basetoken_form_work_04 = exports.basetoken_form_work_03 = exports.basetoken_form_work_02 = exports.basetoken_form_work_01 = exports.basetoken_form_work = exports.basetoken_form_quiet_04 = exports.basetoken_form_quiet_03 = exports.basetoken_form_quiet_02 = exports.basetoken_form_quiet_01 = exports.basetoken_form_quiet = exports.basetoken_form_peace_04 = exports.basetoken_form_peace_03 = exports.basetoken_form_peace_02 = exports.basetoken_form_peace_01 = exports.basetoken_form_peace = exports.basetoken_form_moon_04 = void 0;
exports.basetoken_template_center_card_bg = exports.basetoken_table_details_process_bg = exports.basetoken_table_bg_yellow_04 = exports.basetoken_table_bg_yellow_03 = exports.basetoken_table_bg_yellow_02 = exports.basetoken_table_bg_yellow_01 = exports.basetoken_table_bg_wathet_04 = exports.basetoken_table_bg_wathet_03 = exports.basetoken_table_bg_wathet_02 = exports.basetoken_table_bg_wathet_01 = exports.basetoken_table_bg_turquoise_04 = exports.basetoken_table_bg_turquoise_03 = exports.basetoken_table_bg_turquoise_02 = void 0;
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
exports.basetoken_AI_instruction_hover = 'rgba(255, 255, 255, 0.1)';
exports.basetoken_AI_instruction_normal = 'rgba(255, 255, 255, 0.05)';
exports.basetoken_AI_instruction_press = 'rgba(255, 255, 255, 0.15)';
exports.basetoken_container_bg = '#0a0a0a';
exports.basetoken_dashboard_sidebar_bg_overlay = '#262626';
exports.basetoken_dashboard_slicer_tag_bg_hover = 'rgba(10, 10, 10, 0.3)';
exports.basetoken_dashboard_slicer_tag_bg_normal = 'rgba(10, 10, 10, 0.2)';
exports.basetoken_form_award = '#909fb1';
exports.basetoken_form_award_01 = 'rgba(144, 159, 177, 0.64)';
exports.basetoken_form_award_02 = 'rgba(144, 159, 177, 0.4)';
exports.basetoken_form_award_03 = 'rgba(144, 159, 177, 0.24)';
exports.basetoken_form_award_04 = 'rgba(144, 159, 177, 0.08)';
exports.basetoken_form_desert = '#aba69e';
exports.basetoken_form_desert_01 = 'rgba(171, 166, 158, 0.64)';
exports.basetoken_form_desert_02 = 'rgba(171, 166, 158, 0.4)';
exports.basetoken_form_desert_03 = 'rgba(171, 166, 158, 0.24)';
exports.basetoken_form_desert_04 = 'rgba(171, 166, 158, 0.08)';
exports.basetoken_form_entertainment = '#afb2a4';
exports.basetoken_form_entertainment_01 = 'rgba(175, 178, 164, 0.64)';
exports.basetoken_form_entertainment_02 = 'rgba(175, 178, 164, 0.4)';
exports.basetoken_form_entertainment_03 = 'rgba(175, 178, 164, 0.24)';
exports.basetoken_form_entertainment_04 = 'rgba(175, 178, 164, 0.08)';
exports.basetoken_form_flower = '#acbac1';
exports.basetoken_form_flower_01 = 'rgba(172, 186, 193, 0.64)';
exports.basetoken_form_flower_02 = 'rgba(172, 186, 193, 0.4)';
exports.basetoken_form_flower_03 = 'rgba(172, 186, 193, 0.24)';
exports.basetoken_form_flower_04 = 'rgba(172, 186, 193, 0.08)';
exports.basetoken_form_fog = '#adb1ab';
exports.basetoken_form_fog_01 = 'rgba(173, 177, 171, 0.64)';
exports.basetoken_form_fog_02 = 'rgba(173, 177, 171, 0.4)';
exports.basetoken_form_fog_03 = 'rgba(173, 177, 171, 0.24)';
exports.basetoken_form_fog_04 = 'rgba(173, 177, 171, 0.08)';
exports.basetoken_form_goal = '#a39281';
exports.basetoken_form_goal_01 = 'rgba(163, 146, 129, 0.64)';
exports.basetoken_form_goal_02 = 'rgba(163, 146, 129, 0.4)';
exports.basetoken_form_goal_03 = 'rgba(163, 146, 129, 0.24)';
exports.basetoken_form_goal_04 = 'rgba(163, 146, 129, 0.08)';
exports.basetoken_form_header_bg = 'rgba(10, 10, 10, 0.4)';
exports.basetoken_form_header_bg_hover = 'rgba(10, 10, 10, 0.5)';
exports.basetoken_form_header_bg_press = 'rgba(10, 10, 10, 0.6)';
exports.basetoken_form_hourglass = '#bab4a5';
exports.basetoken_form_hourglass_01 = 'rgba(186, 180, 165, 0.64)';
exports.basetoken_form_hourglass_02 = 'rgba(186, 180, 165, 0.4)';
exports.basetoken_form_hourglass_03 = 'rgba(186, 180, 165, 0.24)';
exports.basetoken_form_hourglass_04 = 'rgba(186, 180, 165, 0.08)';
exports.basetoken_form_input_bg = 'rgba(68, 68, 68, 0.08)';
exports.basetoken_form_moon = '#b6bbca';
exports.basetoken_form_moon_01 = 'rgba(182, 187, 202, 0.64)';
exports.basetoken_form_moon_02 = 'rgba(182, 187, 202, 0.4)';
exports.basetoken_form_moon_03 = 'rgba(182, 187, 202, 0.24)';
exports.basetoken_form_moon_04 = 'rgba(182, 187, 202, 0.08)';
exports.basetoken_form_peace = '#9997b1';
exports.basetoken_form_peace_01 = 'rgba(153, 151, 177, 0.64)';
exports.basetoken_form_peace_02 = 'rgba(153, 151, 177, 0.4)';
exports.basetoken_form_peace_03 = 'rgba(153, 151, 177, 0.24)';
exports.basetoken_form_peace_04 = 'rgba(153, 151, 177, 0.08)';
exports.basetoken_form_quiet = '#a6aeb9';
exports.basetoken_form_quiet_01 = 'rgba(166, 174, 185, 0.64)';
exports.basetoken_form_quiet_02 = 'rgba(166, 174, 185, 0.4)';
exports.basetoken_form_quiet_03 = 'rgba(166, 174, 185, 0.24)';
exports.basetoken_form_quiet_04 = 'rgba(166, 174, 185, 0.08)';
exports.basetoken_form_work = '#beb9b1';
exports.basetoken_form_work_01 = 'rgba(190, 185, 177, 0.64)';
exports.basetoken_form_work_02 = 'rgba(190, 185, 177, 0.4)';
exports.basetoken_form_work_03 = 'rgba(190, 185, 177, 0.24)';
exports.basetoken_form_work_04 = 'rgba(190, 185, 177, 0.08)';
exports.basetoken_home_create_new_bg = '#373737';
exports.basetoken_table_bg_blue_01 = '#152340';
exports.basetoken_table_bg_blue_02 = '#172f5e';
exports.basetoken_table_bg_blue_03 = '#193a7c';
exports.basetoken_table_bg_blue_04 = '#2655b6';
exports.basetoken_table_bg_carmine_01 = '#321827';
exports.basetoken_table_bg_carmine_02 = '#461b34';
exports.basetoken_table_bg_carmine_03 = '#65284b';
exports.basetoken_table_bg_carmine_04 = '#883564';
exports.basetoken_table_bg_green_01 = '#0e2b0a';
exports.basetoken_table_bg_green_02 = '#173b12';
exports.basetoken_table_bg_green_03 = '#21511a';
exports.basetoken_table_bg_green_04 = '#275e20';
exports.basetoken_table_bg_lime_01 = '#212702';
exports.basetoken_table_bg_lime_02 = '#303804';
exports.basetoken_table_bg_lime_03 = '#404c06';
exports.basetoken_table_bg_lime_04 = '#4d5a0c';
exports.basetoken_table_bg_neutral_01 = '#292929';
exports.basetoken_table_bg_neutral_02 = '#373737';
exports.basetoken_table_bg_neutral_03 = '#4c4c4c';
exports.basetoken_table_bg_neutral_04 = '#565656';
exports.basetoken_table_bg_orange_01 = '#33210b';
exports.basetoken_table_bg_orange_02 = '#4a2b10';
exports.basetoken_table_bg_orange_03 = '#683a12';
exports.basetoken_table_bg_orange_04 = '#7f4519';
exports.basetoken_table_bg_purple_01 = '#281943';
exports.basetoken_table_bg_purple_02 = '#381f61';
exports.basetoken_table_bg_purple_03 = '#492688';
exports.basetoken_table_bg_purple_04 = '#5c33a4';
exports.basetoken_table_bg_red_01 = '#3d1a19';
exports.basetoken_table_bg_red_02 = '#591f1d';
exports.basetoken_table_bg_red_03 = '#7b2524';
exports.basetoken_table_bg_red_04 = '#852e2c';
exports.basetoken_table_bg_turquoise_01 = '#132926';
exports.basetoken_table_bg_turquoise_02 = '#173b36';
exports.basetoken_table_bg_turquoise_03 = '#1d4e47';
exports.basetoken_table_bg_turquoise_04 = '#245e57';
exports.basetoken_table_bg_wathet_01 = '#152830';
exports.basetoken_table_bg_wathet_02 = '#103647';
exports.basetoken_table_bg_wathet_03 = '#164359';
exports.basetoken_table_bg_wathet_04 = '#135a78';
exports.basetoken_table_bg_yellow_01 = '#30250a';
exports.basetoken_table_bg_yellow_02 = '#473409';
exports.basetoken_table_bg_yellow_03 = '#63470f';
exports.basetoken_table_bg_yellow_04 = '#7f5d19';
exports.basetoken_table_details_process_bg = 'rgba(255, 255, 255, 0.03)';
exports.basetoken_template_center_card_bg = '#2f2f2f';
