// @byted-nodex/logger-core/lib/transport/console_transport.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const transport_1 = __importDefault(require("./transport"));
class ConsoleTransport extends transport_1.default {
    constructor(options = {}) {
        super(options);
    }
    write(msg) {
        var _a;
        if (typeof ((_a = process.stdout) === null || _a === void 0 ? void 0 : _a.write) === 'function') {
            process.stdout.write(msg);
        }
        else {
            console.warn(`Unsupport operating environment, process.stdout must be a tty or stream. \n' +
          "Now logger's console output just use console.log().`);
            console.log(msg);
        }
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    log(formattedData, _data) {
        this.write(formattedData);
    }
    reload() {
        //
    }
    close() {
        //
    }
    flush() {
        //
    }
    flushSync() {
        //
    }
}
exports.default = ConsoleTransport;
