// @byted-nodex/logger-core/lib/transport/file_transport.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const sonic_boom_1 = __importDefault(require("sonic-boom"));
const cron_1 = require("cron");
const dayjs_1 = __importDefault(require("dayjs"));
const transport_1 = __importDefault(require("./transport"));
class FileTransport extends transport_1.default {
    constructor(options = {}) {
        super(options);
        this.file = options.file || path_1.default.join(process.cwd(), './log/default.log');
        const dirname = path_1.default.dirname(this.file);
        if (!fs_1.default.existsSync(dirname)) {
            fs_1.default.mkdirSync(dirname, {
                recursive: true,
            });
        }
        this.sonic = new sonic_boom_1.default({
            dest: this.file,
            minLength: 0,
            sync: false,
            mkdir: false,
        });
        this.sonic.setMaxListeners(1000);
        this.job = new cron_1.CronJob('0 0 * * * *', () => {
            this.rotate();
        }, null, true);
    }
    rotate() {
        const oldPath = this.file;
        const newPath = `${this.file}.${(0, dayjs_1.default)(new Date())
            .subtract(1, 'hour')
            .format('YYYY-MM-DD_HH:mm:ss')}`;
        if (!fs_1.default.existsSync(oldPath)) {
            this.reload();
            return;
        }
        try {
            if (!fs_1.default.existsSync(newPath)) {
                fs_1.default.renameSync(oldPath, newPath);
            }
            this.reload();
        }
        catch (err) {
            console.log(err);
            this.reload();
        }
    }
    write(message) {
        this.sonic.write(message);
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    log(formattedData, _data) {
        this.write(formattedData);
    }
    reload() {
        this.sonic.reopen(this.file);
    }
    close() {
        this.sonic.end();
        this.job.stop();
    }
    flush() {
        this.sonic.flush();
    }
    flushSync() {
        this.sonic.flushSync();
    }
}
exports.default = FileTransport;
