// @byted-nodex/logger-core/lib/transport/transport.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const level_1 = require("../level");
/**
 * Transport is an output channel of the log that can be output to a file, stdout/stderr or network.
 * A logger can configure multiple Transports to meet a variety of complex needs
 */
class Transport {
    constructor(options = {}) {
        const constructorName = this.constructor.name;
        this.name =
            options.name || constructorName.slice(0, 1).toLowerCase() + constructorName.slice(1);
        this.level = options.level;
        this.enabled = true;
    }
    setLevel(level) {
        this.level = level;
    }
    getLevel() {
        return this.level;
    }
    /**
     * enable transport
     */
    enable() {
        this.enabled = true;
    }
    /**
     * disable transport
     */
    disable() {
        this.enabled = false;
    }
    /**
     * should output log or not
     */
    shouldLog(level) {
        if (!this.enabled || !this.level) {
            return false;
        }
        return level_1.LevelValues[level] >= level_1.LevelValues[this.level];
    }
}
exports.default = Transport;
