// @universe-design/react/src/skeleton/index.tsx
import clsx from 'clsx';
import React from 'react';
import { __SCREENSHOT__ } from '../util';
import { useCssScope } from '../config-provider';

import './style';

export interface SkeletonProps {
  /**
   * 加载完毕后实际内容
   */
  children?: React.ReactNode;
  /**
   * 类名透传
   */
  className?: string;
  /**
   * 加载中，为 true 时显示骨架屏，反之显示 `children` 内容
   * @defaultValue true
   */
  loading?: boolean;
  /**
   * 是否开启动画
   * @defaultValue true
   */
  animation?: boolean;
  /**
   * 占位元素
   */
  placeholder?: React.ReactNode;
  /**
   * 样式属性透传
   */
  style?: React.CSSProperties;
}

export const Skeleton: React.FC<SkeletonProps> = (props: SkeletonProps) => {
  const { className, loading = true, placeholder, animation = !__SCREENSHOT__, style, children } = props;

  const cssScope = useCssScope('skeleton');

  if (loading) {
    return (
      <div
        className={clsx(className, {
          [cssScope()]: true,
          [cssScope('-animation')]: animation,
        })}
        style={style}
      >
        {placeholder}
      </div>
    );
  }
  return <>{children}</>;
};

export { SkeletonContent, SkeletonAvatar, SkeletonParagraph } from './ShapeItem';

export type { SkeletonContentProps, SkeletonAvatarProps, SkeletonParagraphProps } from './ShapeItem';
