// @universe-design/react/src/skeleton/ShapeItem.tsx
import clsx from 'clsx';
import merge from 'lodash/merge';
import React from 'react';
import { useCssScope } from '../config-provider';

import './style';

type SkeletonTypeName = 'avatar' | 'content';

interface ShapeItemProps {
  className?: string;
  typeName?: SkeletonTypeName;
  style?: React.CSSProperties;
}

export interface SkeletonContentProps {
  /**
   * 类名透传
   */
  className?: string;
  /**
   * 样式属性透传
   */
  style?: React.CSSProperties;
  /**
   * 矩形块宽度
   * @defaultValue 100%
   */
  width?: React.CSSProperties['width'];
  /**
   * 矩形块高度
   * @defaultValue 14px
   */
  height?: React.CSSProperties['height'];
}

export interface SkeletonAvatarProps {
  /**
   * 类名透传
   */
  className?: string;
  /**
   * 样式属性透传
   */
  style?: React.CSSProperties;
  /**
   * 圆形块直径
   * @defaultValue 32
   */
  size?: number;
}

export interface SkeletonParagraphProps {
  /**
   * 类名透传
   */
  className?: string;
  /**
   * 样式属性透传
   */
  style?: React.CSSProperties;
  /**
   * 段落中矩形块行数，矩形块样式为`SkeletonContent`的默认样式
   * @defaultValue 3
   */
  rows?: number;
}

const ShapeItem: React.FC<ShapeItemProps> = props => {
  const { className, typeName, style } = props;

  const cssScope = useCssScope('skeleton__item');

  return (
    <div className={clsx([cssScope(), cssScope(`-${typeName}`), className])} style={style}>
      <div className={cssScope('__animation-layer')} />
    </div>
  );
};

export const SkeletonContent = (props: SkeletonContentProps) => {
  const { className, style, width, height } = props;
  // skip undefined
  const mergedStyle = merge(style, { width, height });
  return <ShapeItem typeName="content" className={className} style={mergedStyle} />;
};

export const SkeletonAvatar = (props: SkeletonAvatarProps) => {
  const { className, style, size } = props;
  // skip undefined
  const mergedStyle = merge(style, { width: size, height: size });
  return <ShapeItem typeName="avatar" className={className} style={mergedStyle} />;
};

export const SkeletonParagraph = (props: SkeletonParagraphProps) => {
  const { className, style, rows = 3 } = props;
  const cssScope = useCssScope('skeleton');

  return (
    <div className={clsx([cssScope('__paragraph'), className])} style={style}>
      {[...Array(rows)].map((_, index) => (
        <SkeletonContent key={index} />
      ))}
    </div>
  );
};
