// @universe-design/react/src/illustration/stories/homeless.stories.tsx
import React, { useState } from 'react';
import * as illustration from '@universe-design/illustration-react';
import { Illustration, Button, Space, Slider } from '../..';
import { EmptyNeutralNoContent } from '@universe-design/illustration-react';

export default {
  title: 'Components/Illustration/临时示例', // 注意：这里需要修改成自己对应 Component name。
  component: Illustration,
};

export const AllIllustration = () => (
  <Space wrap>
    {Object.entries(illustration)
      .filter(([name]) => !/illustrationTheme|(Light|Dark$)/.test(name))
      .map(([name, C]) => {
        const CC = C as (
          props: React.SVGProps<SVGSVGElement>,
          // eslint-disable-next-line no-undef
        ) => JSX.Element;
        return (
          <div key={name}>
            {name}
            <CC />
          </div>
        );
      })}
  </Space>
);

export const Width = () => {
  const [maxWidth, setMaxWidth] = useState<number>(250);
  return (
    <Space style={{ width: '100%' }} direction="vertical" size="large">
      <Illustration image={<illustration.EmptyNeutralNoAccess />} title="没有权限访问" description="请登录后尝试访问">
        <Button type="primary">申请权限</Button>
        <Button>返回主页</Button>
      </Illustration>
      <Illustration
        style={{ width: 250 }}
        image={<illustration.EmptyNeutralNoAccess />}
        title="无内容"
        description="副标题：负反馈为空多用于数据、内容为空，或给用户带来负面操作的示意"
      >
        <Button style={{ width: 225 }} type="primary">
          次要操作 用于按钮上下叠放模式
        </Button>
        <Button style={{ width: 225 }}>次要操作 用于下叠放模式</Button>
      </Illustration>
      <div style={{ margin: '0 auto', width: '50%' }}>
        <span>最大宽度：</span>
        <Slider max={1000} min={250} value={maxWidth} onChange={setMaxWidth} />
      </div>
      <Illustration
        style={{ maxWidth }}
        image={<illustration.EmptyNeutralNoAccess />}
        title={`自定义最大宽度：${maxWidth}px`}
        description="副标题：负反馈为空多用于数据、内容为空，或给用户带来负面操作的示意"
      >
        <Button style={{ width: 225 }} type="primary">
          次要操作 用于按钮上下叠放模式
        </Button>
        <Button style={{ width: 225 }}>次要操作 用于下叠放模式</Button>
      </Illustration>
    </Space>
  );
};

export const ShowAll = () => (
  <Space style={{ width: '100%' }} direction="vertical" size="large">
    <Space align="start">
      <Illustration
        image={<EmptyNeutralNoContent />}
        title="正反馈为空: 无需操作"
        description="副标题：正反馈多用于用户无需操做，正向、无影响的为空表现"
      >
        <Button type="primary">主要按钮</Button>
        <Button>次要操作</Button>
      </Illustration>
      <Illustration
        image={<EmptyNeutralNoContent />}
        title="正反馈为空: 无需操作"
        description="副标题：正反馈多用于用户无需操做，正向、无影响的为空表现"
      >
        <Button type="primary">主要按钮</Button>
      </Illustration>
      <Illustration
        image={<EmptyNeutralNoContent />}
        description="副标题：正反馈多用于用户无需操做，正向、无影响的为空表现"
      >
        <Button type="primary">主要按钮</Button>
      </Illustration>
      <Illustration
        image={<EmptyNeutralNoContent />}
        description="副标题：正反馈多用于用户无需操做，正向、无影响的为空表现"
      />

      <Illustration image={<EmptyNeutralNoContent />} description={false} />
    </Space>
  </Space>
);

export { default as Loading2 } from '../demo/loading2';
