// @universe-design/react/src/illustration/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import CBasic from '../demo/basic';
import CCustomize from '../demo/customize';
import CDescription from '../demo/description';
import CLoading from '../demo/loading';
import CStyle from '../demo/style';
if (!(CBasic as Story).storyName) {
  (CBasic as Story).storyName = '0. 基础用法';
  (CBasic as Story).parameters = {
    docs: {
      description: {
        story: '简单的用法。该用法和直接使用 `@universe-design/illustration-react` 库里面的组件效果相同。\n\n',
      },
    },
  };
}
if (!(CCustomize as Story).storyName) {
  (CCustomize as Story).storyName = '2. 自定义图片';
  (CCustomize as Story).parameters = {
    docs: {
      description: {
        story: '通过`image`自定义图标，通过`imageStyle`自定义图片样式。\n\n',
      },
    },
  };
}
if (!(CDescription as Story).storyName) {
  (CDescription as Story).storyName = '1. 自定义描述内容';
  (CDescription as Story).parameters = {
    docs: {
      description: {
        story: '通过`description`自定义描述内容，也可以通过 `title` 属性添加标题。\n\n',
      },
    },
  };
}
if (!(CLoading as Story).storyName) {
  (CLoading as Story).storyName = '5. Loading';
  (CLoading as Story).parameters = {
    docs: {
      description: {
        story: '可以引用`EmptyPositiveLoading` 实现 loading 动画。\n',
      },
    },
  };
}
if (!(CStyle as Story).storyName) {
  (CStyle as Story).storyName = '3. 自定义样式';
  (CStyle as Story).parameters = {
    docs: {
      description: {
        story: '通过`style`或`className`自定义样式。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Illustration/官网示例',
  tags: ['autodocs'],
};

export { CBasic, CCustomize, CDescription, CLoading, CStyle };
