// @universe-design/react/lib/date-picker/panels/base/Body.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _clsx = _interopRequireDefault(require("clsx"));
var React = _interopRequireWildcard(require("react"));
var _hooks = require("../../../config-provider/hooks");
var _cell = _interopRequireDefault(require("../../cell"));
var _DateCalendarUIContext = require("../DateCalendar/DateCalendarUIContext");
function Body(props) {
  const {
    className,
    cellClassName,
    rowNum,
    colNum,
    baseDate,
    autoHeight,
    getCellDate = date => date,
    dateRender = () => null,
    getCellStatus,
    // headerCells,
    children
  } = props;
  const prefixCls = (0, _hooks.useCssScope)('picker')();
  const bodyPrefixCls = `${prefixCls}-panel-body`;
  const {
    onDateMouseEnter,
    onDateMouseLeave,
    onSelect,
    onContentMouseEnter,
    onContentMouseLeave
  } = (0, _DateCalendarUIContext.useDateCalendarUIContext)();
  const rows = [];
  for (let rowIndex = 0; rowIndex < rowNum; rowIndex++) {
    const isRowSelected = [...new Array(colNum)].some((_, colIndex) => getCellStatus(getCellDate(baseDate, rowIndex * colNum + colIndex)).isSelected);
    rows.push(/*#__PURE__*/React.createElement("div", {
      className: (0, _clsx.default)(`${bodyPrefixCls}-row`, isRowSelected && `${bodyPrefixCls}-row--selected`),
      key: rowIndex
    }, [...new Array(colNum)].map((_, colIndex) => {
      const offset = rowIndex * colNum + colIndex;
      const cellDate = getCellDate(baseDate, offset);
      const cellStatus = getCellStatus(cellDate);
      const {
        isDisabled
      } = cellStatus;
      const preCellDate = getCellDate(baseDate, offset - 1);
      const nextCellDate = getCellDate(baseDate, offset + 1);
      const prevCellStatus = getCellStatus(preCellDate);
      const nextCellStatus = getCellStatus(nextCellDate);
      return /*#__PURE__*/React.createElement(React.Fragment, {
        key: colIndex
      }, /*#__PURE__*/React.createElement(_cell.default, (0, _extends2.default)({
        onClick: () => {
          if (!isDisabled) {
            onSelect(cellDate);
          }
        },
        onMouseEnter: () => {
          if (!isDisabled) {
            onDateMouseEnter(cellDate);
          }
        },
        onMouseLeave: () => {
          if (!isDisabled) {
            onDateMouseLeave(cellDate);
          }
        },
        prevCellStatus: prevCellStatus,
        nextCellStatus: nextCellStatus,
        className: cellClassName,
        colIndex: colIndex,
        colNum: colNum
      }, cellStatus), dateRender(cellDate)));
    })));
  }
  return /*#__PURE__*/React.createElement("div", {
    className: (0, _clsx.default)(bodyPrefixCls, autoHeight && `${prefixCls}-auto-height`, className)
  }, children, /*#__PURE__*/React.createElement("div", {
    className: (0, _clsx.default)(`${bodyPrefixCls}-content`, 'notranslate'),
    onMouseEnter: onContentMouseEnter,
    onMouseLeave: onContentMouseLeave
  }, rows));
}
var _default = exports.default = Body;
