// @universe-design/react/lib/date-picker/panels/base/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getYearStart = exports.getYearRange = exports.getPrev = exports.getNext = exports.getHigherLevelMode = exports.YEAR_COUNT = void 0;
exports.isInRange = isInRange;
const YEAR_COUNT = exports.YEAR_COUNT = 20;
const getYearStart = date => {
  const newDate = date.clone();
  return newDate.year(Math.floor(newDate.year() / YEAR_COUNT) * YEAR_COUNT);
};
exports.getYearStart = getYearStart;
const getYearRange = date => {
  const start = getYearStart(date).year();
  return [start, start + YEAR_COUNT - 1].join(' - ');
};
exports.getYearRange = getYearRange;
function generatorDateByMode(date, mode, offset) {
  const newDate = date.clone();
  switch (mode) {
    case 'date':
      return newDate.add(offset, 'month');
    case 'month':
      return newDate.add(offset, 'year');
    case 'year':
      // return newDate.add(offset * YEAR_COUNT, 'year');
      return getYearStart(newDate).add(offset * YEAR_COUNT, 'year');
    default:
      return newDate;
  }
}
const getNext = (date, mode) => generatorDateByMode(date, mode, 1);
exports.getNext = getNext;
const getPrev = (date, mode) => generatorDateByMode(date, mode, -1);
exports.getPrev = getPrev;
const getHigherLevelMode = mode => {
  const levels = ['date', 'month', 'year'];
  const index = levels.findIndex(m => m === mode);
  if (index < levels.length - 1) {
    return levels[index + 1];
  }
  return null;
};
exports.getHigherLevelMode = getHigherLevelMode;
function isInRange(date, range, mode) {
  return date.isSameOrAfter(range[0], mode) && date.isSameOrBefore(range[1], mode);
}
