// @universe-design/react/src/affix/stories/homeless.stories.tsx
import React, { useState } from 'react';
import { Button, Affix } from '../..';

export default {
  title: 'Components/Affix/临时示例',
  component: Affix,
};

export const DifferentStates = () => (
  <div style={{ height: '500vh', paddingTop: 100 }}>
    <div>
      <Affix offsetTop={20}>
        <Button type="primary">affix top 20 px</Button>
      </Affix>
      <div style={{ height: 8 }} />
      <Affix offsetTop={60}>
        <Button type="primary">affix top 60 px</Button>
      </Affix>
    </div>
    <div style={{ marginTop: 1500 }}>
      <Affix offsetBottom={120}>
        <Button type="primary">affix bottom 120 px</Button>
      </Affix>
      <Affix offsetBottom={80}>
        <Button style={{ marginTop: 8 }} type="primary">
          affix bottom 80 px
        </Button>
      </Affix>
    </div>
  </div>
);

export const AffixWithContainer = () => {
  const [container, setContainer] = useState<HTMLDivElement | null>(null);
  return (
    <div style={{}}>
      <div
        style={{
          height: 200,
          overflowY: 'auto',
          backgroundColor: 'pink',
          width: 300,
        }}
        ref={setContainer}
      >
        <div
          style={{
            paddingTop: 20,
            height: 500,
          }}
        >
          <Affix target={() => container as HTMLDivElement}>
            <Button type="primary">吸顶！！！</Button>
          </Affix>
        </div>
      </div>
    </div>
  );
};
