// @universe-design/react/src/affix/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import CBasic from '../demo/basic';
import CBottom from '../demo/bottom';
import COnChange from '../demo/on-change';
import CTarget from '../demo/target';
import CTop from '../demo/top';
if (!(CBasic as Story).storyName) {
  (CBasic as Story).storyName = '0. 基础用法';
  (CBasic as Story).parameters = {
    docs: {
      description: {
        story: '当页面滚动元素不可见时，元素固定在页面最顶部。\n\n',
      },
    },
  };
}
if (!(CBottom as Story).storyName) {
  (CBottom as Story).storyName = '2. 底部固定';
  (CBottom as Story).parameters = {
    docs: {
      description: {
        story: '当页面滚动或浏览器窗口改变时，元素向下滚动到距底部一定距离时固定。\n\n',
      },
    },
  };
}
if (!(COnChange as Story).storyName) {
  (COnChange as Story).storyName = '4. 监听固定状态改变';
  (COnChange as Story).parameters = {
    docs: {
      description: {
        story: '当固定状态发生改变时，会触发事件。\n\n',
      },
    },
  };
}
if (!(CTarget as Story).storyName) {
  (CTarget as Story).storyName = '4. 滚动容器';
  (CTarget as Story).parameters = {
    docs: {
      description: {
        story: '通过`target`设置需要监听其滚动事件的元素。\n\n',
      },
    },
  };
}
if (!(CTop as Story).storyName) {
  (CTop as Story).storyName = '1. 顶部固定';
  (CTop as Story).parameters = {
    docs: {
      description: {
        story: '当页面滚动或浏览器窗口改变时，元素向上滚动到距顶部一定距离时固定。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Affix/官网示例',
  tags: ['autodocs'],
};

export { CBasic, CBottom, COnChange, CTarget, CTop };
