// @universe-design/react/lib/shared/components/picker/utils/date.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.generateFormat = void 0;
exports.isEqual = isEqual;
exports.isNullEqual = isNullEqual;
exports.isSameDate = isSameDate;
exports.isSameTime = isSameTime;
function isNullEqual(value1, value2) {
  if (!value1 && !value2) {
    return true;
  }
  if (!value1 || !value2) {
    return false;
  }
  return undefined;
}
function isSameDate(generateConfig, date1, date2) {
  const equal = isNullEqual(date1, date2);
  if (typeof equal === 'boolean') {
    return equal;
  }
  return generateConfig.getYear(date1) === generateConfig.getYear(date2) && generateConfig.getMonth(date1) === generateConfig.getMonth(date2) && generateConfig.getDate(date1) === generateConfig.getDate(date2);
}
function isSameTime(generateConfig, time1, time2) {
  const equal = isNullEqual(time1, time2);
  if (typeof equal === 'boolean') {
    return equal;
  }
  // time1 或者 time2 为 null 的时候返回 false
  if (!time1 || !time2) {
    return false;
  }
  return generateConfig.getHour(time1) === generateConfig.getHour(time2) && generateConfig.getMinute(time1) === generateConfig.getMinute(time2) && generateConfig.getSecond(time1) === generateConfig.getSecond(time2);
}
function isEqual(generateConfig, value1, value2) {
  return isSameDate(generateConfig, value1, value2) && isSameTime(generateConfig, value1, value2);
}
const generateFormat = exports.generateFormat = {
  is12Hour: format => format.includes('h') || format.includes('a') || format.includes('A'),
  showSecond: format => format.includes('s'),
  showMinute: format => format.includes('m'),
  showHour: format => format.includes('h') || format.includes('H') || format.includes('k')
};
