// @universe-design/react-utilities/lib-commonjs/components/createInOutTransition.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createInOutTransition = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _reactTransitionGroup = require("react-transition-group");
var _useStyleSheetOnce = require("../hooks/useStyleSheetOnce");
var _index = require("../util/index");
var _util = require("../util");
const useTransitionStyles = props => {
  const {
    name,
    timeout,
    useCssScope
  } = props;
  // Timeout
  let timeoutAppear = 0;
  let timeoutEnter = 0;
  let timeoutExit = 0;
  if (timeout) {
    if (typeof timeout === 'number') {
      timeoutAppear = timeout;
      timeoutEnter = timeout;
      timeoutExit = timeout;
    } else {
      var _timeout$appear, _timeout$enter, _timeout$exit;
      timeoutAppear = (_timeout$appear = timeout.appear) !== null && _timeout$appear !== void 0 ? _timeout$appear : 0;
      timeoutEnter = (_timeout$enter = timeout.enter) !== null && _timeout$enter !== void 0 ? _timeout$enter : 0;
      timeoutExit = (_timeout$exit = timeout.exit) !== null && _timeout$exit !== void 0 ? _timeout$exit : 0;
    }
  }

  // ClassName
  const classNames = useCssScope(name)(`-${timeoutAppear}-${timeoutEnter}-${timeoutExit}`);
  return {
    classNames,
    timeout: {
      appear: timeoutAppear,
      enter: timeoutEnter,
      exit: timeoutExit
    }
  };
};
/**
 * @public
 */
const createInOutTransition = config => {
  const SomeInOutTransition = props => {
    const {
      transitions: transitionsProp,
      extraProps,
      ...rest
    } = props;
    const {
      timeout: timeoutProp
    } = props;
    const name = (() => {
      if ((0, _index.isFunction)(config.name)) {
        if (!extraProps) {
          throw new Error('`config.name` is function, but extraProps is undefined.');
        }
        return config.name(extraProps);
      }
      return config.name;
    })();
    const {
      classNames,
      timeout
    } = useTransitionStyles({
      timeout: timeoutProp,
      name,
      useCssScope: config.useCssScope
    });
    const selectors = (0, _util.prefixToSelector)(classNames);
    const styles = (() => {
      if ((0, _index.isFunction)(config.styles)) {
        if (!extraProps) {
          throw new Error('`config.styles` is function, but extraProps is undefined.');
        }
        return config.styles(extraProps);
      }
      return config.styles;
    })();

    // Inject stylesheet
    (0, _useStyleSheetOnce.useStyleSheetOnce)(`
  ${selectors}-appear {
    ${styles.appear}
  }
  ${selectors}-enter {
    ${styles.enter}
  }
  ${selectors}-exit-done {
    ${styles.exitDone}
  }
  
  ${selectors}-enter-active {
    ${styles.enterActive}
    transition: ${config.transitions.enterActive(timeout.enter)}${transitionsProp !== null && transitionsProp !== void 0 && transitionsProp.enterActive ? `, ${transitionsProp.enterActive(timeout.enter)}` : ''};
  }
  ${selectors}-appear-active {
    ${styles.appearActive}
    transition: ${config.transitions.appearActive(timeout.appear)}${transitionsProp !== null && transitionsProp !== void 0 && transitionsProp.appearActive ? `, ${transitionsProp.appearActive(timeout.appear)}` : ''};
  }

  ${selectors}-appear-done {
    ${styles.appearDone}
  }
  ${selectors}-enter-done {
    ${styles.enterDone}
  }
  ${selectors}-exit {
    ${styles.exit}
  }
  
  ${selectors}-exit-active {
    ${styles.exitActive}
    transition: ${config.transitions.exitActive(timeout.exit)}${transitionsProp !== null && transitionsProp !== void 0 && transitionsProp.exitActive ? `, ${transitionsProp.exitActive(timeout.exit)}` : ''};
  }`, {
      atLeastOne: true
    });
    return /*#__PURE__*/_react.default.createElement(_reactTransitionGroup.CSSTransition, (0, _extends2.default)({}, rest, {
      classNames: classNames,
      timeout: timeout
    }));
  };
  if (process.env.NODE_ENV !== 'production') {
    SomeInOutTransition.displayName = `InOutTransition(${config.name})`;
  }
  return SomeInOutTransition;
};
exports.createInOutTransition = createInOutTransition;
