// @universe-design/styles/ts/global/transition-timing-function.ts
// autogen by ../../scripts/gen-ts-from-json.ts
export const transition_timing_function_ease_base_in = 'cubic-bezier(0.9, 0, 0.3, 0.7)';
export const transition_timing_function_ease_base_out = 'cubic-bezier(0.7, 0.3, 0.1, 1)';
export const transition_timing_function_ease_in = 'cubic-bezier(0.55, 0.055, 0.675, 0.19)';
export const transition_timing_function_ease_in_back = 'cubic-bezier(0.71, -0.46, 0.88, 0.6)';
export const transition_timing_function_ease_in_circ = 'cubic-bezier(0.6, 0.04, 0.98, 0.34)';
export const transition_timing_function_ease_in_out = 'cubic-bezier(0.645, 0.045, 0.355, 1)';
export const transition_timing_function_ease_in_out_back = 'cubic-bezier(0.71, -0.46, 0.29, 1.46)';
export const transition_timing_function_ease_in_out_circ = 'cubic-bezier(0.78, 0.14, 0.15, 0.86)';
export const transition_timing_function_ease_in_out_quint = 'cubic-bezier(0.86, 0, 0.07, 1)';
export const transition_timing_function_ease_in_quint = 'cubic-bezier(0.755, 0.05, 0.855, 0.06)';
export const transition_timing_function_ease_out = 'cubic-bezier(0.215, 0.61, 0.355, 1)';
export const transition_timing_function_ease_out_back = 'cubic-bezier(0.12, 0.4, 0.29, 1.46)';
export const transition_timing_function_ease_out_circ = 'cubic-bezier(0.08, 0.82, 0.17, 1)';
export const transition_timing_function_ease_out_quint = 'cubic-bezier(0.23, 1, 0.32, 1)';
export const transition_timing_function_standard = 'cubic-bezier(0.34, 0.69, 0.1, 1)';
