// @universe-design/react/src/shared/components/trigger/index.tsx
import type { Ref } from 'react';
import React, { forwardRef, useCallback, useImperativeHandle, useMemo, useRef, useState } from 'react';
import type { TriggerProps as RcTriggerProps } from '@universe-design/rc-trigger';
import RcTrigger from '@universe-design/rc-trigger';
import merge from 'lodash/merge';
import cloneDeep from 'lodash/cloneDeep';
import pick from 'lodash/pick';
import { useControlledState } from '@universe-design/react-utilities';
import { useClassPrefix } from '../../../config-provider/hooks';
import type {
  BuildInPlacements,
  RcTriggerType,
  TriggerActionType,
  TriggerHideActionType,
  TriggerShowActionType,
  UDAlignType,
  UdTriggerRef,
  UdTRiggerArrow,
} from './types';
import type { LiteralUnion } from '../../../util';
import { isFunction, __SCREENSHOT__, devWarning, warning } from '../../../util';
import { useIsomorphicLayoutEffect as useLayoutEffect } from '../../../hooks';
import type { CSSMotionProps } from '@universe-design/rc-motion';
import { useTriggerDisabled } from './modifiers/useTriggerDisabled';
import clsx from 'clsx';
import Popup from './Popup';
import { useTriggerObserver } from './modifiers/useTriggerObserver';
import composeRefs from '../../../util/composeRefs';
import { isOverflowed as defaultIsOverflow } from '../../../util/dom';
import { usePopupPortalContainer } from '@universe-design/react-provider';

/**
 * @docgen
 * @public
 */
export interface UDTriggerProps
  extends Omit<
    RcTriggerProps,
    | 'popupMotion'
    | 'maskMotion'
    | 'action'
    | 'showAction'
    | 'hideAction'
    | 'builtinPlacements'
    | 'popupAlign'
    | 'destroyPopupOnHide'
    | 'autoDestroy'
    | 'popupVisible'
    | 'defaultPopupVisible'
    | 'onPopupVisibleChange'
    | 'children'
  > {
  children: React.ReactElement | ((props: { visible: boolean }) => React.ReactElement);
  /**
   * 是否禁用 trigger
   */
  disabled?: boolean;
  /**
   * 箭头配置信息
   */
  arrow?: UdTRiggerArrow;
  /**
   * 触发动作
   */
  action?: TriggerActionType | TriggerActionType[];
  /**
   * 显示弹层的触发动作
   */
  showAction?: TriggerShowActionType[];
  /**
   * 隐藏弹层的触发动作
   */
  hideAction?: TriggerHideActionType[];
  /**
   * 自定义动画或使用 UD 提供的内置动画
   */
  popupMotion?: LiteralUnion<
    'slide-up' | 'slide-down' | "'slide-left" | 'slide-right' | 'tooltip-zoom',
    string | CSSMotionProps
  >;
  /**
   * 内置的一些位置和对齐规则
   */
  builtinPlacements?: BuildInPlacements;
  /**
   * 触发打开动作的延迟，单位为毫秒
   * @defaultValue 100
   */
  mouseEnterDelay?: number;
  /**
   * 触发关闭动作的延迟，单位为毫秒
   * @defaultValue 100
   */
  mouseLeaveDelay?: number;
  /**
   * @deprecated popupAlign 属性已废弃，若使用 `FloatingUIPopperSidecar` 请使用 `shift`、`flip`、`offset`、`rootBoundary` 属性代替
   */
  popupAlign?: UDAlignType;
  /**
   * 修饰符，提供 trigger 的内建增强功能
   * @defaultValue { forcePopupAlign: true, popupOverflowAuto: true }
   */
  modifiers?: {
    /**
     * 所有滚动父级的滚动事件、 window 的 resize 事件、trigger 和 popup resize 事件触发时强制进行 popup 的对齐
     */
    forcePopupAlign?: boolean;
    /**
     * popup 在不进行 adjust align 的时候可以内部滚动
     */
    popupOverflowAuto?: boolean;
  };
  /**
   * 关闭后是否销毁内容节点，当 `keepParent` 为 `false` 时销毁父容器
   * @defaultValue `{ keepParent: false }`
   */
  destroyOnHide?: boolean | { keepParent?: boolean };
  /**
   * 弹层是否可见的受控值
   */
  visible?: boolean;
  /**
   * 弹层是否可见的默认值
   * @defaultValue false
   */
  defaultVisible?: boolean;
  /**
   * 弹层显示隐藏的回调
   */
  onVisibleChange?: (visible: boolean) => void;
  /**
   * 弹层打开时触发元素的 `className` 属性透传
   */
  openClassName?: string;
  /**
   * 获取 ud-trigger ref
   * @internal
   */
  udTriggerRef?: Ref<UdTriggerRef>;
  /**
   * 是否开启溢出显示文字提示
   * @defaultValue false
   */
  checkOverflow?: boolean | ((ele: Element) => boolean);
  /**
   * screenshot 时是否始终打开 popper
   * @internal
   * @defaultValue false
   */
  __screenshotAlwaysOpenEnabled?: boolean;
  // 控制是否阻止 popup 点击事件冒泡，设置 click 为 true 时阻止事件传播
  popupStopPropagation?: { click?: boolean };
}

export function getAlignFromPlacement(
  builtinPlacements: BuildInPlacements,
  placementStr: string,
  align: UDAlignType,
): UDAlignType {
  const baseAlign = builtinPlacements[placementStr] || {};
  return merge(cloneDeep(baseAlign), align);
}

export const defaults = {
  mouseEnterDelay: 100,
  mouseLeaveDelay: 100,
  destroyOnHide: { keepParent: false },
  popupAlign: {},
};

export const UDTrigger = forwardRef<RcTriggerType, UDTriggerProps>((props, ref) => {
  const rootPrefixCls = useClassPrefix();
  const {
    prefixCls = `${rootPrefixCls}-trigger`,
    disabled = false,
    popupMotion = 'slide-up',
    arrow: arrowProps,
    mouseEnterDelay = defaults.mouseEnterDelay,
    mouseLeaveDelay = defaults.mouseLeaveDelay,
    popupPlacement,
    popupAlign: popupAlignProp = defaults.popupAlign,
    builtinPlacements,
    modifiers = {},
    destroyOnHide = defaults.destroyOnHide,
    visible: visibleProp,
    checkOverflow = false,
    defaultVisible = false,
    onVisibleChange,
    openClassName = `${prefixCls}-open`,
    className,
    udTriggerRef,
    getPopupContainer: getPopupContainerProp,
    popupClassName,
    children,
    __screenshotAlwaysOpenEnabled = false,
    popupStopPropagation,
    ...restProps
  } = props;

  const { popupOverflowAuto: enablePopupOverflowAuto = false, forcePopupAlign: enableforcePopupAlign = false } =
    modifiers;

  // getPopupContainer
  const ctxGetPopupContainer = usePopupPortalContainer();

  // 获取用于对齐的 popupAlign 配置，逻辑从 rc-trigger 提升
  const popupAlign = useMemo(() => {
    if (popupPlacement && builtinPlacements) {
      return getAlignFromPlacement(builtinPlacements, popupPlacement, popupAlignProp);
    }
    return popupAlignProp;
  }, [builtinPlacements, popupAlignProp, popupPlacement]);

  // visible
  const [visible, setVisible] = useControlledState(defaultVisible, visibleProp, { onChange: onVisibleChange });

  const triggerRef = useRef<RcTriggerType | null>(null);

  const checkVisible = useCallback(
    (_visible: boolean) => {
      if (_visible && checkOverflow) {
        const child = (triggerRef.current as any)?.getRootDomNode();
        const overflowChecker = typeof checkOverflow === 'function' ? checkOverflow : defaultIsOverflow;
        if (!child || !overflowChecker(child)) {
          setVisible(false);
          return;
        }
      }
      setVisible(_visible);
    },
    [checkOverflow, setVisible],
  );

  __SCREENSHOT__ &&
    __screenshotAlwaysOpenEnabled &&
    // eslint-disable-next-line react-hooks/rules-of-hooks
    React.useEffect(() => {
      (triggerRef.current as any)?.forcePopupAlign();
      checkVisible(true);
    });

  // destroy
  let destroyPopupOnHide = false;
  let autoDestroy = false;
  if (typeof destroyOnHide === 'boolean') {
    destroyPopupOnHide = destroyOnHide;
  } else if (destroyOnHide && typeof destroyOnHide === 'object') {
    const { keepParent } = destroyOnHide;
    destroyPopupOnHide = keepParent === true;
    autoDestroy = keepParent === false;
  }

  // 预处理 props
  const rcTriggerProps: RcTriggerProps = {
    ...restProps,
    children: isFunction(children) ? children({ visible }) : children,
    popupClassName: clsx(popupClassName, mouseLeaveDelay === 0 && `${prefixCls}-no-entry`),
    // forceRender 的优先级大于 destroyPopupOnHide
    // 俩都传了 true 的时候, 相当于 hide 后不会销毁
    destroyPopupOnHide,
    autoDestroy,
    mouseEnterDelay: mouseEnterDelay / 1000,
    mouseLeaveDelay: mouseLeaveDelay / 1000,
    prefixCls,
    popupMotion: { motionName: `${rootPrefixCls}-${popupMotion}` },
    popupAlign,
    builtinPlacements,
    popupPlacement,
    popupVisible: visible,
    /**
     * 这里现在有点问题，rc-trigger onPopupVisibleChange 在 visible false 的时候不能触发，所以 checkOverflow 开启时，如果子元素从 overflow 变化到不 overflow，鼠标移出 trigger 区域，子元素再从不 overflow 变成 overflow，tooltip 会再次出现
     */
    onPopupVisibleChange: checkVisible,
    getPopupContainer: getPopupContainerProp ?? ctxGetPopupContainer,
  };

  // 禁用 trigger
  const [newProps] = useTriggerDisabled(rcTriggerProps, { disabled });

  useImperativeHandle(
    udTriggerRef,
    () => ({
      visible,
      setVisible,
      forcePopupAlign: () => {
        (triggerRef?.current as any)?.forcePopupAlign();
      },
    }),
    [visible, setVisible],
  );

  const [readyObserver, setReadyObserver] = useState(false);
  const startObserver = useCallback(() => setReadyObserver(true), []);

  const [childSize, popupSize, viewPointRect] = useTriggerObserver({
    triggerRef,
    popupAlign,
    readyObserver,
    enableforcePopupAlign,
    enablePopupOverflowAuto,
  });

  useLayoutEffect(() => {
    if (!visible && readyObserver) {
      setReadyObserver(false);
    }
  }, [visible, readyObserver]);

  const composeTrigger = useMemo(() => composeRefs(triggerRef, ref), [triggerRef, ref]);

  if (disabled) {
    return <>{rcTriggerProps.children}</>;
  }

  // TODO 临时逻辑，处理 menu eventKey 透传
  warning(
    !('eventKey' in newProps),
    devWarning(
      'Trigger',
      // eslint-disable-next-line max-len
      'The `eventKey` property is detected, maybe you wrap the popper elements such as <Tooltip /> or <Popover /> outside the <MenuItem /> or <SubMenu />, it is recommended to move it to the inside',
    ),
  );
  newProps.children = React.cloneElement(newProps.children, pick(newProps, ['eventKey']));

  return (
    <RcTrigger
      {...newProps}
      popup={
        <Popup
          enablePopupOverflowAuto={enablePopupOverflowAuto}
          childSize={childSize}
          popupSize={popupSize}
          viewPointRect={viewPointRect}
          prefixCls={newProps.prefixCls}
          popup={newProps.popup}
          triggerRef={triggerRef}
          popupAlign={popupAlign}
          arrow={arrowProps}
          popupStopPropagation={popupStopPropagation}
        />
      }
      className={clsx(className, visible && openClassName)}
      popupVisible={visible}
      onPopupAlign={startObserver}
      ref={composeTrigger}
    />
  );
});
if (__DEV__) {
  UDTrigger.displayName = 'UDTrigger';
}

export * from './types';
