// @universe-design/react/src/shared/components/trigger/index.stories.tsx
import React, { useEffect, useMemo, useRef, useState } from 'react';
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import { useDraggable } from 'use-draggable';
import { UDTrigger } from './index';
import '../../../globalStyle';
import './index.less';
import type { BuildInPlacements } from '@universe-design/rc-trigger';
import { Switch } from '../../..';
import { updateRef } from '../../../util/composeRefs';

export default {
  title: 'Shared Components/UDTrigger', // 注意：这里需要修改成自己对应 Component name。
  component: UDTrigger,
};

const getPlacements = ({
  alwaysByViewport,
  adjustY,
  adjustX,
}: {
  alwaysByViewport: boolean;
  adjustY: boolean;
  adjustX: boolean;
}): BuildInPlacements => {
  const defaultAlign = {
    targetOffset: [0, 0],
    ignoreShake: true,
  };

  const autoAdjustOverflow = {
    adjustX,
    adjustY,
    alwaysByViewport,
  };

  const offset = 4 + 7;

  return {
    left: {
      points: ['cr', 'cl'],
      overflow: autoAdjustOverflow,
      offset: [-offset, 0],
      ...defaultAlign,
    },
    right: {
      points: ['cl', 'cr'],
      overflow: autoAdjustOverflow,
      offset: [offset, 0],
      ...defaultAlign,
    },
    top: {
      points: ['bc', 'tc'],
      overflow: autoAdjustOverflow,
      offset: [0, -offset],
      ...defaultAlign,
    },
    bottom: {
      points: ['tc', 'bc'],
      overflow: autoAdjustOverflow,
      offset: [0, offset],
      ...defaultAlign,
    },
    topLeft: {
      points: ['bl', 'tl'],
      overflow: autoAdjustOverflow,
      offset: [0, -offset],
      ...defaultAlign,
    },
    leftTop: {
      points: ['tr', 'tl'],
      overflow: autoAdjustOverflow,
      offset: [-offset, 0],
      ...defaultAlign,
    },
    topRight: {
      points: ['br', 'tr'],
      overflow: autoAdjustOverflow,
      offset: [0, -offset],
      ...defaultAlign,
    },
    rightTop: {
      points: ['tl', 'tr'],
      overflow: autoAdjustOverflow,
      offset: [offset, 0],
      ...defaultAlign,
    },
    bottomRight: {
      points: ['tr', 'br'],
      overflow: autoAdjustOverflow,
      offset: [0, offset],
      ...defaultAlign,
    },
    rightBottom: {
      points: ['bl', 'br'],
      overflow: autoAdjustOverflow,
      offset: [offset, 0],
      ...defaultAlign,
    },
    bottomLeft: {
      points: ['tl', 'bl'],
      overflow: autoAdjustOverflow,
      offset: [0, offset],
      ...defaultAlign,
    },
    leftBottom: {
      points: ['br', 'bl'],
      overflow: autoAdjustOverflow,
      offset: [-offset, 0],
      ...defaultAlign,
    },
  };
};

const menu = (
  <div style={{ width: 120 }}>
    {Array(10)
      .fill(0)
      .map((o, i) => (
        <div style={{ border: '1px solid black' }} key={i}>
          {i}
        </div>
      ))}
  </div>
);

export const CCMOverflowScroll = () => {
  const alignContainerRef = useRef<HTMLDivElement>(null);
  const { targetRef } = useDraggable({ controlStyle: true });
  const [alwaysByViewport, setAlwaysByViewport] = useState(true);
  const [adjustY, setAdjustY] = useState(true);
  const [adjustX, setAdjustX] = useState(true);

  useEffect(() => {
    alignContainerRef.current?.scrollTo({
      // eslint-disable-next-line @typescript-eslint/no-magic-numbers
      top: 600 - 450 / 2,
    });
  }, []);

  return (
    <>
      <Switch checked={alwaysByViewport} onChange={setAlwaysByViewport} />
      alwaysByViewport
      <Switch checked={adjustY} onChange={setAdjustY} />
      adjustY
      <Switch checked={adjustX} onChange={setAdjustX} />
      adjustX
      <div
        ref={alignContainerRef}
        style={{
          width: 400,
          height: 450,
          overflow: 'auto',
          border: '1px red solid',
          position: 'relative',
        }}
      >
        <div style={{ width: 1, height: 600 }} />
        <UDTrigger
          getPopupContainer={() => alignContainerRef.current as HTMLElement}
          // popupVisible={true}
          popupPlacement="bottom"
          arrow={{
            size: 7,
          }}
          builtinPlacements={getPlacements({
            alwaysByViewport,
            adjustY,
            adjustX,
          })}
          action={['click']}
          popup={menu}
        >
          <button ref={targetRef} type="button">
            确定
          </button>
        </UDTrigger>
        <div style={{ width: 1, height: 600 }} />
      </div>
    </>
  );
};

export const IMAlignContainer = () => {
  const alignContainerRef = useRef<HTMLDivElement>(null);
  const { targetRef } = useDraggable({ controlStyle: true });
  return (
    <>
      <div
        style={{
          background: 'red',
          width: 400,
          height: 600,
          overflow: 'auto',
          border: '1px red solid',
          position: 'relative',
        }}
      >
        <div
          ref={alignContainerRef}
          style={{
            background: 'green',
            position: 'absolute',
            left: 0,
            right: 100,
            top: 0,
            bottom: 0,
            overflow: 'auto',
          }}
        />
      </div>
      <UDTrigger
        getPopupContainer={() => alignContainerRef.current as HTMLElement}
        // popupVisible={true}
        popupPlacement="top"
        builtinPlacements={useMemo(
          () =>
            getPlacements({
              alwaysByViewport: true,
              adjustY: true,
              adjustX: true,
            }),
          [],
        )}
        action={['click']}
        popup={menu}
      >
        <button
          ref={node => {
            updateRef(targetRef, node);
          }}
          type="button"
        >
          确定
        </button>
      </UDTrigger>
    </>
  );
};

export const NonReplaced = () => (
  <UDTrigger
    popupPlacement="top"
    builtinPlacements={useMemo(
      () =>
        getPlacements({
          alwaysByViewport: true,
          adjustY: true,
          adjustX: false,
        }),
      [],
    )}
    action={['click']}
    popup={menu}
    arrow={{ size: 7 }}
  >
    <span>确定</span>
  </UDTrigger>
);
