// @universe-design/react/src/shared/components/trigger/Popup.tsx
import type React from 'react';
import { cloneElement, isValidElement, useMemo } from 'react';
import type { TriggerProps as RcTriggerProps } from '@universe-design/rc-trigger';
import { isFunction, isNonNullable } from '../../../util';
import { composeEventHandlers } from '../../../util/function';
import { useTriggerArrow } from './modifiers/useTriggerArrow';
import type { UDAlignType, RcTriggerType, RectInfo, UdTRiggerArrow } from './types';
import type { SizeInfo } from '../../../hooks';

interface PopupProps {
  childSize: SizeInfo;
  popupSize: SizeInfo;
  viewPointRect: RectInfo | null;
  popupAlign: UDAlignType;
  popup: RcTriggerProps['popup'];
  prefixCls?: string;
  triggerRef: React.RefObject<RcTriggerType | null>;
  arrow?: UdTRiggerArrow;
  enablePopupOverflowAuto?: boolean;
  popupStopPropagation?: { click?: boolean };
}

export default function Popup(props: PopupProps) {
  const {
    popupAlign,
    arrow,
    prefixCls,
    childSize,
    popupSize,
    viewPointRect,
    enablePopupOverflowAuto = false,
    popupStopPropagation: { click: popupClickStopPropagation = false } = {},
  } = props;

  let { popup: popupElement } = props;

  const popupContentStyle = useMemo(() => {
    const style: React.CSSProperties = {};
    if (!enablePopupOverflowAuto) {
      return undefined;
    }

    // 如果两个方向都要翻转就不做 maxWidth 和 maxHeight 的计算了
    if (popupAlign.overflow?.adjustX && popupAlign.overflow?.adjustY) {
      return undefined;
    }
    if (!viewPointRect) {
      return undefined;
    }
    style.overflow = 'auto';
    if (!popupAlign.overflow?.adjustX) {
      if (childSize.right < popupSize.left) {
        style.maxWidth = viewPointRect.right - popupSize.left;
      } else if (childSize.left > popupSize.right) {
        style.maxWidth = popupSize.right - viewPointRect.left;
      }
    }
    if (!popupAlign.overflow?.adjustY) {
      if (childSize.bottom < popupSize.top) {
        style.maxHeight = viewPointRect.bottom - popupSize.top;
      } else if (childSize.top > popupSize.bottom) {
        style.maxHeight = popupSize.bottom - viewPointRect.top;
      }
    }
    return style;
  }, [popupAlign, viewPointRect, popupSize, childSize, enablePopupOverflowAuto]);

  popupElement = isFunction(popupElement) ? popupElement() : popupElement;
  popupElement = isValidElement<{ style?: React.CSSProperties; onClick?: (e: React.MouseEvent) => void }>(popupElement)
    ? cloneElement(popupElement, {
        style: isNonNullable(popupContentStyle)
          ? {
              ...popupContentStyle,
              ...popupElement.props.style,
            }
          : {},
        onClick: composeEventHandlers((e: React.MouseEvent) => {
          if (popupClickStopPropagation) {
            e.stopPropagation();
          }
        }, popupElement.props.onClick),
      })
    : popupElement;

  popupElement = useTriggerArrow({
    childSize,
    popupSize,
    arrow,
    popup: popupElement,
    prefixCls,
  });

  return popupElement as React.ReactElement;
}
