// @universe-design/react/src/shared/components/trigger/types.ts
import type { TriggerProps as RcTriggerProps } from '@universe-design/rc-trigger';
import type { AlignType } from '@universe-design/rc-trigger/es/interface';
import type * as React from 'react';

export type TriggerActionType = 'click' | 'contextMenu' | 'hover' | 'focus';
export type TriggerShowActionType = 'click' | 'contextMenu' | 'mouseEnter' | 'focus';
export type TriggerHideActionType = 'click' | 'mouseLeave' | 'blur';

export type RcTriggerType = React.Component<RcTriggerProps, any, any>;

export interface RcTriggerPropsWithRef extends RcTriggerProps {
  ref?: React.Ref<RcTriggerType>;
}

/**
 * 对齐方式
 * @public
 */
export interface UDAlignType extends AlignType {
  /**
   * 弹窗超出边界的处理方法
   */
  overflow?: {
    adjustX?: boolean | number;
    adjustY?: boolean | number;
    alwaysByViewport?: boolean;
  };
}

export type BuildInPlacements = Record<string, UDAlignType>;

/**
 * @public
 */
export interface UdTriggerRef {
  /**
   * popup 是否可见
   */
  visible: boolean;
  /**
   * 设置是否显示弹层
   */
  setVisible: React.Dispatch<boolean>;
  /**
   * 重新计算弹层位置
   */
  forcePopupAlign: () => void;
}

export interface RectInfo {
  top: number;
  left: number;
  right: number;
  bottom: number;
}

/**
 * @public
 */
export interface UdTRiggerArrow {
  /**
   * 箭头的尺寸，等于宽度值的一半
   */
  size?: number;
  /**
   * 箭头安全边距是指箭头出现在弹层的极限位置离弹层真实的边界有多远
   */
  safeMarginX?: number;
  safeMarginY?: number;
  /**
   * 自定义内容节点
   */
  content?: ((props: { placement: 'left' | 'right' | 'top' | 'bottom' }) => React.ReactNode) | React.ReactNode;
}
