// @universe-design/react/es/date-picker/utils/date.js
import moment from 'moment';
export function isNullEqual(value1, value2) {
  if (!value1 && !value2) {
    return true;
  }
  if (!value1 || !value2) {
    return false;
  }
  return undefined;
}
export function isEqual(value1, value2, mode) {
  let showTime = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : false;
  const equal = isNullEqual(value1, value2);
  if (typeof equal === 'boolean') {
    return equal;
  }
  return value1 === null || value1 === void 0 ? void 0 : value1.isSame(value2, showTime ? undefined : mode);
}
export function isValidRangeValue(range) {
  const date1 = range[0];
  const date2 = range[1];
  if (!date1 && !date2 || date1 && date2 && date1.isSameOrBefore(date2)) {
    return true;
  }
  return false;
}
export function isEqualRange(range1, range2, mode) {
  const equal = isNullEqual(range1, range2);
  if (typeof equal === 'boolean') {
    return equal;
  }
  return isEqual(range1[0], range2[0], mode) && isEqual(range1[1], range2[1], mode);
}
export function getWeekStartDate(value, startOfWeek) {
  const startMonth = value.clone().startOf('month');
  const weekFirstDay = startMonth.day();
  return startMonth.clone().subtract((weekFirstDay - startOfWeek + 7) % 7, 'day');
}
export function mergeDateAndTime(date, time) {
  if (!time) {
    return date;
  }
  return date.clone().hours(time.hours()).minutes(time.minutes()).seconds(time.seconds());
}

/**
 * 当a与b均为undefined时 此时moment(a)与moment(b)是否相等依赖于执行的的时间
 */
export function isSame(a, b) {
  if (a === undefined && b === undefined) {
    return true;
  }
  if (a !== undefined && b !== undefined) {
    return moment(a).isSame(b);
  }
  return false;
}
