// @byted-service/zti-helper/lib/spiffeid.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpiffeidID = void 0;
const url_1 = require("url");
class SpiffeidID {
    constructor(id, path) {
        this.td = id;
        this.path = path;
    }
    string() {
        /* istanbul ignore next */
        if (this.isZero()) {
            return '';
        }
        return `spiffe://${this.td.name}/${this.path}`;
    }
    isZero() {
        return !this.td.name;
    }
    static fromString(str) {
        if (!str) {
            throw new Error(`unable to parse: str is empty`);
        }
        const uri = new url_1.URL(str);
        if (uri.protocol !== 'spiffe:') {
            throw new Error('invalid scheme');
        }
        if (uri.username) {
            throw new Error('user info is not allowed');
        }
        /* istanbul ignore next */
        if (!uri.host) {
            throw new Error('trust domain is empty');
        }
        if (uri.port) {
            throw new Error('port is not allowed');
        }
        if (uri.searchParams.toString() != '') {
            throw new Error('query is not allowed');
        }
        return new SpiffeidID({ name: uri.host.toLowerCase() }, uri.pathname);
    }
}
exports.SpiffeidID = SpiffeidID;
