// @byted-service/zti-helper/lib/utill.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getRegionFromSystemEnv = exports.decodeIdentityFromTokenString = exports.validateTokenAlgorithm = exports.validateWithLeeway = exports.parseInsecure = exports.IS_SCM = void 0;
const tslib_1 = require("tslib");
const env_1 = require("@byted-service/env");
const jwt = tslib_1.__importStar(require("jsonwebtoken"));
const spiffeid_1 = require("./spiffeid");
const zeroTrustIdentity_1 = require("./zeroTrustIdentity");
const consts_1 = require("./consts");
exports.IS_SCM = !!process.env.BUILD_REPO_NAME;
function parseInsecure(token, audience) {
    const result = jwt.decode(token, { complete: true });
    if (!result || typeof result.payload === 'string') {
        throw new Error('decode token failed token: ' + token);
    }
    const { payload } = result;
    validateTokenAlgorithm(result);
    /* istanbul ignore next */
    if (!payload.sub) {
        throw new Error('token missing subject claim');
    }
    /* istanbul ignore next */
    if (!payload.exp) {
        throw new Error('token missing exp claim');
    }
    const spiffeID = spiffeid_1.SpiffeidID.fromString(payload.sub);
    validateWithLeeway(payload, { aud: audience, time: Date.now() }, 60 * 1000);
    const svid = {
        token,
        audience: typeof payload.aud === 'object' ? payload.aud : [payload.aud],
        expiry: payload.exp,
        claims: Object.assign({}, payload),
        id: spiffeID,
    };
    return svid;
}
exports.parseInsecure = parseInsecure;
function validateWithLeeway(payload, expected, leeway) {
    if (expected.iss && payload.iss !== expected.iss) {
        throw new Error('validation failed, invalid issuer claim (issuer)');
    }
    if (expected.sub && payload.sub !== expected.sub) {
        throw new Error('validation failed, invalid subject claim (subject)');
    }
    if (expected.id && payload.id !== expected.id) {
        throw new Error('validation failed, invalid ID claim (ID)');
    }
    if (expected.aud && expected.aud.length > 0) {
        if (typeof expected.aud !== typeof payload.aud) {
            throw new Error('validation failed, invalid audience claim (audience)');
        }
        if (typeof expected.aud === 'string') {
            if (expected.aud !== payload.aud) {
                throw new Error('validation failed, invalid audience claim (audience)');
            }
        }
        else {
            const aud = expected.aud;
            const payloadAud = payload.aud;
            payloadAud.forEach(a => {
                if (!aud.includes(a)) {
                    throw new Error('validation failed, invalid audience claim (audience)');
                }
            });
        }
    }
    if (expected.time) {
        if (payload.nbf && expected.time + leeway < payload.nbf * 1000) {
            throw new Error('validation failed, token not valid yet (nbf)');
        }
        if (payload.exp && expected.time - leeway > payload.exp * 1000) {
            throw new Error('validation failed, token is expired (exp)');
        }
        if (payload.iat && expected.time < payload.iat * 1000) {
            throw new Error('validation field, token issued in the future (iat)');
        }
    }
}
exports.validateWithLeeway = validateWithLeeway;
function validateTokenAlgorithm(result) {
    // TODO 这里 Golang 是判断 headers 的长度这里需要跟 golang 的同学对一下
    if (typeof result.header !== 'object') {
        throw new Error(`expected a single token header; got ${result.header}`);
    }
    const alg = result.header.alg;
    if (!(alg === 'RS256' ||
        alg === 'RS384' ||
        alg === 'RS512' ||
        alg === 'ES256' ||
        alg === 'ES384' ||
        alg === 'ES512' ||
        alg === 'PS256' ||
        alg === 'PS384' ||
        alg === 'PS512')) {
        throw new Error(`unsupported token signature algorithm ${alg}`);
    }
}
exports.validateTokenAlgorithm = validateTokenAlgorithm;
function decodeIdentityFromTokenString(tokenString, token) {
    token = token || jwt.decode(tokenString, { complete: true });
    if (!token || typeof token.payload === 'string') {
        throw new Error('decode token failed, token: ' + tokenString);
    }
    const { payload } = token;
    const spiffeID = spiffeid_1.SpiffeidID.fromString(payload.sub);
    const zti = new zeroTrustIdentity_1.ZeroTrustIdentity(spiffeID, payload.exp || 0);
    zti.validateZTI();
    return zti;
}
exports.decodeIdentityFromTokenString = decodeIdentityFromTokenString;
function getRegionFromSystemEnv() {
    const ztiEnv = process.env[consts_1.ZTI_ENV];
    if (ztiEnv) {
        return ztiEnv.toUpperCase();
    }
    /* istanbul ignore next */
    return (0, env_1.getRegion)();
}
exports.getRegionFromSystemEnv = getRegionFromSystemEnv;
