// @universe-design/react-ai/lib-commonjs/shared/components/LottieAnimation.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _lottieWeb = _interopRequireDefault(require("lottie-web"));
var _reactUtilities = require("@universe-design/react-utilities");
var _default = exports.default = /*#__PURE__*/React.forwardRef((props, ref) => {
  const {
    animationData,
    onSegmentPlayEnd,
    lottieAnimationRef: lottieAnimationRefProp,
    ...htmlProps
  } = props;
  const containerRef = React.useRef(null);
  const mergedContainerRef = (0, _reactUtilities.useMergedRefs)(ref, containerRef);
  const lottieAnimationRef = React.useRef();
  const mergedLottieAnimationRef = (0, _reactUtilities.useMergedRefs)(lottieAnimationRefProp, lottieAnimationRef);
  (0, _reactUtilities.useEffectOnce)(() => {
    if (containerRef.current) {
      const animation = _lottieWeb.default.loadAnimation({
        animationData,
        container: containerRef.current,
        // the dom element that will contain the animation
        renderer: 'svg',
        loop: false,
        autoplay: false
      });
      mergedLottieAnimationRef && (0, _reactUtilities.updateRef)(mergedLottieAnimationRef, animation);
      return () => {
        animation.destroy();
        mergedLottieAnimationRef && (0, _reactUtilities.updateRef)(mergedLottieAnimationRef, null);
      };
    }
  });
  React.useEffect(() => {
    const animation = lottieAnimationRef.current;
    if (!animation || !onSegmentPlayEnd) {
      return;
    }
    animation.addEventListener('complete', onSegmentPlayEnd);
    return () =>
    // animation 已经提前销毁，lottie 内部 removeEventListener 的时候没有判空
    animation._cbs && animation.removeEventListener('complete', onSegmentPlayEnd);
  }, [onSegmentPlayEnd]);
  return /*#__PURE__*/React.createElement("div", (0, _extends2.default)({
    ref: mergedContainerRef
  }, htmlProps));
});
