// @universe-design/react-ai/lib-commonjs/shared/components/ThinkingAnimationBase.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.starting = exports.looping = exports.ending = exports.ThinkingAnimationBase = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _LottieAnimation = _interopRequireDefault(require("./LottieAnimation"));
var _reactUtilities = require("@universe-design/react-utilities");
const initial = 'initial';
const starting = exports.starting = 'starting';
const looping = exports.looping = 'looping';
const ending = exports.ending = 'ending';
const ThinkingAnimationBase = exports.ThinkingAnimationBase = /*#__PURE__*/React.forwardRef((props, ref) => {
  const {
    thinking,
    segments,
    ...htmlProps
  } = props;
  const animationRef = React.useRef(null);
  const statusRef = React.useRef(initial);
  const onSegmentPlayEnd = (0, _reactUtilities.useEventCallback)(() => {
    if (statusRef.current === ending && thinking) {
      var _animationRef$current;
      (_animationRef$current = animationRef.current) === null || _animationRef$current === void 0 ? void 0 : _animationRef$current.playSegments(segments[starting], false);
      statusRef.current = starting;
      return;
    }
    if (statusRef.current === ending && !thinking) {
      statusRef.current = initial;
      return;
    }
    if (statusRef.current === starting && thinking) {
      var _animationRef$current2;
      (_animationRef$current2 = animationRef.current) === null || _animationRef$current2 === void 0 ? void 0 : _animationRef$current2.playSegments(segments[looping], false);
      statusRef.current = looping;
      return;
    }
    if (statusRef.current === starting && !thinking) {
      var _animationRef$current3;
      (_animationRef$current3 = animationRef.current) === null || _animationRef$current3 === void 0 ? void 0 : _animationRef$current3.playSegments(segments[ending], false);
      statusRef.current = ending;
      return;
    }
    if (statusRef.current === looping && thinking) {
      var _animationRef$current4;
      (_animationRef$current4 = animationRef.current) === null || _animationRef$current4 === void 0 ? void 0 : _animationRef$current4.playSegments(segments[looping], false);
      return;
    }
    if (statusRef.current === looping && !thinking) {
      var _animationRef$current5;
      (_animationRef$current5 = animationRef.current) === null || _animationRef$current5 === void 0 ? void 0 : _animationRef$current5.playSegments(segments[ending], false);
      statusRef.current = ending;
    }
  });
  React.useEffect(() => {
    // start play
    if (thinking && statusRef.current === initial) {
      var _animationRef$current6;
      (_animationRef$current6 = animationRef.current) === null || _animationRef$current6 === void 0 ? void 0 : _animationRef$current6.playSegments(segments[starting], true);
      statusRef.current = starting;
    }
  }, [segments, thinking]);
  return /*#__PURE__*/React.createElement(_LottieAnimation.default, (0, _extends2.default)({
    ref: ref,
    lottieAnimationRef: animationRef,
    onSegmentPlayEnd: onSegmentPlayEnd
  }, htmlProps));
});
