// @universe-design/rc-table/lib/FixedHolder/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _typeof = require("@babel/runtime/helpers/typeof").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var React = _interopRequireWildcard(require("react"));
var _classnames = _interopRequireDefault(require("classnames"));
var _ref2 = require("@universe-design/rc-util/lib/ref");
var _ColGroup = _interopRequireDefault(require("../ColGroup"));
var _TableContext = require("../context/TableContext");
var _excluded = ["className", "noData", "columns", "flattenColumns", "colWidths", "columCount", "stickyOffsets", "direction", "fixHeader", "stickyTopOffset", "stickyBottomOffset", "stickyClassName", "onScroll", "maxContentScroll", "children"];
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function useColumnWidth(colWidths, columCount) {
  return (0, React.useMemo)(function () {
    var cloneColumns = [];
    for (var i = 0; i < columCount; i += 1) {
      var val = colWidths[i];
      if (val !== undefined) {
        cloneColumns[i] = val;
      } else {
        return null;
      }
    }
    return cloneColumns;
  }, [colWidths.join('_'), columCount]);
}
var useIntersectionObserver = function useIntersectionObserver() {
  var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var _options$disabled = options.disabled,
    disabled = _options$disabled === void 0 ? false : _options$disabled;
  var stickyPinnedDetectorRef = React.useRef();
  var _React$useState = React.useState(false),
    _React$useState2 = (0, _slicedToArray2.default)(_React$useState, 2),
    pinned = _React$useState2[0],
    setPinned = _React$useState2[1];
  React.useEffect(function () {
    if (disabled) {
      return;
    }
    if (!window || !window.IntersectionObserver) {
      return;
    }
    var stickyPinnedDetector = stickyPinnedDetectorRef.current;
    if (!stickyPinnedDetector) return;
    var observer = new IntersectionObserver(function (e) {
      if (e[0].intersectionRatio <= 0) {
        setPinned(true);
      } else {
        setPinned(false);
      }
    });
    observer.observe(stickyPinnedDetector);
    return function () {
      observer.disconnect();
    };
  }, [disabled]);
  return [stickyPinnedDetectorRef, pinned];
};
var FixedHolder = /*#__PURE__*/React.forwardRef(function (_ref, ref) {
  var _classNames;
  var className = _ref.className,
    noData = _ref.noData,
    columns = _ref.columns,
    flattenColumns = _ref.flattenColumns,
    colWidths = _ref.colWidths,
    columCount = _ref.columCount,
    stickyOffsets = _ref.stickyOffsets,
    direction = _ref.direction,
    fixHeader = _ref.fixHeader,
    stickyTopOffset = _ref.stickyTopOffset,
    stickyBottomOffset = _ref.stickyBottomOffset,
    stickyClassName = _ref.stickyClassName,
    onScroll = _ref.onScroll,
    maxContentScroll = _ref.maxContentScroll,
    children = _ref.children,
    props = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  var prefixCls = (0, _TableContext.useTableContextSelector)(function (ctx) {
    return ctx.prefixCls;
  });
  var isSticky = (0, _TableContext.useTableContextSelector)(function (ctx) {
    return ctx.isSticky;
  });
  var scrollbarColumnWidth = (0, _TableContext.useTableContextSelector)(function (ctx) {
    return ctx.scrollbarColumnWidth;
  });
  var hasStickyTop = stickyTopOffset !== undefined;
  var hasStickyBottom = stickyBottomOffset !== undefined;
  // Pass wheel to scroll event
  var scrollRef = React.useRef(null);
  var setScrollRef = React.useCallback(function (element) {
    (0, _ref2.fillRef)(ref, element);
    (0, _ref2.fillRef)(scrollRef, element);
  }, [ref]);
  React.useEffect(function () {
    var scrollEl = scrollRef.current;
    if (!scrollEl) return;
    function onWheel(e) {
      var currentTarget = e.currentTarget,
        deltaX = e.deltaX;
      if (deltaX) {
        onScroll({
          currentTarget: currentTarget,
          scrollLeft: currentTarget.scrollLeft + deltaX
        });
        e.preventDefault();
      }
    }
    scrollEl.addEventListener('wheel', onWheel);
    return function () {
      scrollEl.removeEventListener('wheel', onWheel);
    };
  }, [onScroll]);
  // Check if all flattenColumns has width
  var allFlattenColumnsWithWidth = React.useMemo(function () {
    return flattenColumns.every(function (column) {
      return column.width >= 0;
    });
  }, [flattenColumns]);
  // Add scrollbar column
  var lastColumn = flattenColumns[flattenColumns.length - 1];
  var ScrollBarColumn = React.useMemo(function () {
    return {
      fixed: lastColumn ? lastColumn.fixed : null,
      scrollbar: true,
      onHeaderCell: function onHeaderCell() {
        return {
          className: "".concat(prefixCls, "-cell-scrollbar")
        };
      }
    };
  }, [lastColumn, prefixCls]);
  var columnsWithScrollbar = (0, React.useMemo)(function () {
    return scrollbarColumnWidth ? [].concat((0, _toConsumableArray2.default)(columns), [ScrollBarColumn]) : columns;
  }, [scrollbarColumnWidth, columns, ScrollBarColumn]);
  var flattenColumnsWithScrollbar = (0, React.useMemo)(function () {
    return scrollbarColumnWidth ? [].concat((0, _toConsumableArray2.default)(flattenColumns), [ScrollBarColumn]) : flattenColumns;
  }, [scrollbarColumnWidth, flattenColumns, ScrollBarColumn]);
  // 当 stickyHeader pinned 的时候，添加一个特殊的类名
  var _useIntersectionObser = useIntersectionObserver({
      disabled: !isSticky
    }),
    _useIntersectionObser2 = (0, _slicedToArray2.default)(_useIntersectionObser, 2),
    stickyPinnedDetectorTopRef = _useIntersectionObser2[0],
    pinnedTop = _useIntersectionObser2[1];
  var _useIntersectionObser3 = useIntersectionObserver({
      disabled: !isSticky
    }),
    _useIntersectionObser4 = (0, _slicedToArray2.default)(_useIntersectionObser3, 2),
    stickyPinnedDetectorBottomRef = _useIntersectionObser4[0],
    pinnedBottom = _useIntersectionObser4[1];
  // Calculate the sticky offsets
  var headerStickyOffsets = (0, React.useMemo)(function () {
    var right = stickyOffsets.right,
      left = stickyOffsets.left;
    return (0, _objectSpread2.default)((0, _objectSpread2.default)({}, stickyOffsets), {}, {
      left: direction === 'rtl' ? [].concat((0, _toConsumableArray2.default)(left.map(function (width) {
        return width + scrollbarColumnWidth;
      })), [0]) : left,
      right: direction === 'rtl' ? right : [].concat((0, _toConsumableArray2.default)(right.map(function (width) {
        return width + scrollbarColumnWidth;
      })), [0]),
      isSticky: isSticky
    });
  }, [stickyOffsets, direction, isSticky, scrollbarColumnWidth]);
  var mergedColumnWidth = useColumnWidth(colWidths, columCount);
  var stickyPinnedDetectorStyle = {
    height: 1,
    backgroundColor: 'transparent'
  };
  return /*#__PURE__*/React.createElement(React.Fragment, null, hasStickyTop && /*#__PURE__*/React.createElement("div", {
    ref: stickyPinnedDetectorTopRef,
    style: stickyPinnedDetectorStyle
  }), /*#__PURE__*/React.createElement("div", {
    style: (0, _objectSpread2.default)((0, _objectSpread2.default)((0, _objectSpread2.default)({
      overflow: 'hidden'
    }, isSticky ? {
      top: stickyTopOffset,
      bottom: stickyBottomOffset
    } : {}), hasStickyTop && {
      marginTop: pinnedTop ? 0 : -1
    }), hasStickyBottom && {
      marginBottom: pinnedBottom ? 0 : -1
    }),
    ref: setScrollRef,
    className: (0, _classnames.default)(className, (_classNames = {}, (0, _defineProperty2.default)(_classNames, stickyClassName, !!stickyClassName), (0, _defineProperty2.default)(_classNames, "".concat(prefixCls, "__sticky--pinned-top"), pinnedTop), (0, _defineProperty2.default)(_classNames, "".concat(prefixCls, "__sticky--pinned-bottom"), pinnedBottom), _classNames))
  }, /*#__PURE__*/React.createElement("table", {
    style: {
      tableLayout: 'fixed',
      visibility: noData || mergedColumnWidth ? null : 'hidden'
    }
  }, (!noData || !maxContentScroll || allFlattenColumnsWithWidth) && /*#__PURE__*/React.createElement(_ColGroup.default, {
    colWidths: mergedColumnWidth ? [].concat((0, _toConsumableArray2.default)(mergedColumnWidth), [scrollbarColumnWidth]) : [],
    columCount: columCount + 1,
    columns: flattenColumnsWithScrollbar
  }), children((0, _objectSpread2.default)((0, _objectSpread2.default)({}, props), {}, {
    stickyOffsets: headerStickyOffsets,
    columns: columnsWithScrollbar,
    flattenColumns: flattenColumnsWithScrollbar
  })))), hasStickyBottom && /*#__PURE__*/React.createElement("div", {
    ref: stickyPinnedDetectorBottomRef,
    style: stickyPinnedDetectorStyle
  }));
});
FixedHolder.displayName = 'FixedHolder';
var _default = FixedHolder;
exports.default = _default;
