// @universe-design/react/es/shared/components/picker/Comp/panels/time-panel/Column.js
/* eslint-disable @typescript-eslint/no-magic-numbers */

import VList from '@universe-design/rc-virtual-list';
import stylex from '@universe-design/stylex';
import clsx from 'clsx';
import { useCssScope } from '../../../../../../config-provider';
import React, { useRef } from 'react';
import { useEffectOnce, useUpdateEffect } from '../../../../../../hooks';
import { __TEST__ } from '../../../../../../util';
import { ListItem, ListItemTitle } from '../../../../list';
import { ListBox, ListScrollView } from '../../../../list/List';
import { usePickerContextSelector } from '../../../Context';
import { scrollTo } from '../../../utils';
import { isSameTime } from '../../../utils/date';
import { cssTypographyTitle4 } from '../../../../../../stylex';
function Column(props) {
  const {
    units,
    onPanelChange,
    value,
    hideDisabledOptions,
    mode,
    generateConfig,
    ulWidth
  } = props;
  const inDatePicker = usePickerContextSelector(o => o.inDatePicker);
  // 是否被嵌套在 datePicker 中
  const cssScope = useCssScope('time-panel');
  const prefixCls = cssScope();
  const itemCount = inDatePicker ? 264 / 32 : 7.5;
  const listRef = useRef(null);
  const vListRef = useRef(null);
  const liRefs = useRef(new Map());
  const scrollToItem = React.useCallback(function (_value) {
    let duration = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 120;
    const li = liRefs.current.get(_value);
    const list = listRef.current;
    if (li && list) {
      // 第一个元素 margin-top: 1;
      scrollTo(list, li.offsetTop - 1, duration);
    }
    const vList = vListRef.current;
    if (vList) {
      vList.scrollTo({
        key: _value,
        align: 'top'
      });
    }
  }, []);
  useEffectOnce(() => {
    // TODO 这里不用异步会获取不到 li 的尺寸和位置，后面需要看一下
    requestAnimationFrame(() => {
      scrollToItem(Number(value), 0);
    });
  });
  useUpdateEffect(React.useCallback(() => {
    scrollToItem(Number(value));
  }, [scrollToItem, value]));
  const isSameTimeDateType = (time1, time2) => {
    if (mode === 'timeCombine') {
      return time1 !== -1 ? isSameTime(generateConfig, time1, time2) : false;
    } else {
      return false;
    }
  };
  const isSame = (time1, time2) => {
    if (mode === 'timeCombine') {
      return isSameTimeDateType(time1, time2);
    }
    return time1 === time2;
  };
  const renderItem = unit => {
    if (hideDisabledOptions && unit.disabled) {
      return null;
    }
    return /*#__PURE__*/React.createElement(ListItem, {
      key: Number(unit.value),
      ref: element => {
        liRefs.current.set(Number(unit.value), element);
      },
      onClick: () => {
        if (unit.disabled) {
          return;
        }
        onPanelChange && onPanelChange(unit.value);
      },
      disabled: unit.disabled,
      selected: isSame(value, unit.value),
      showCheckmark: mode === 'timeCombine'
    }, /*#__PURE__*/React.createElement(ListItemTitle, null, unit.label));
  };
  return (() => mode === 'timeCombine' ? /*#__PURE__*/React.createElement(VList, {
    data: units !== null && units !== void 0 ? units : [],
    itemKey: unit => Number(unit.value),
    fullHeight: false,
    itemHeight: 32,
    height: 32 * itemCount,
    virtual: !__TEST__,
    ref: vListRef,
    style: {
      width: '100%'
    }
  }, unit => {
    const item = renderItem(unit);
    if (item) {
      return /*#__PURE__*/React.createElement("div", null, item);
    }
  }) : /*#__PURE__*/React.createElement(ListBox, {
    className: `${prefixCls}__column`
  }, inDatePicker && /*#__PURE__*/React.createElement("span", {
    className: clsx(`${prefixCls}__column-title`, stylex(cssTypographyTitle4))
  }, props.title), /*#__PURE__*/React.createElement(ListScrollView, {
    style: {
      width: ulWidth,
      height: itemCount * 32,
      position: 'relative' // ele.scrollTop is offset of closest relative positioned parent.
    },
    className: `${prefixCls}__column-scroll-list`,
    ref: listRef
  }, (units !== null && units !== void 0 ? units : []).map(renderItem), /*#__PURE__*/React.createElement("div", {
    style: {
      paddingBottom: (itemCount - 1) * 32
    }
  }))))();
}
export default Column;
