// @universe-design/react/es/shared/components/picker/Comp/panels/time-panel/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React from 'react';
import { createKeyDownHandler } from '../../../utils';
import NormalTimePanel from './NormalTimePanel';
import CombineTimePanel from './CombineTimePanel';
import { generateFormat } from '../../../utils/date';
const countBoolean = boolList => boolList.filter(bool => bool !== false).length;
export default function TimePanel(props) {
  const {
    generateConfig,
    prefixCls,
    operationRef,
    value,
    format,
    onPanelChange,
    mode = 'time',
    showSecond: showSecondProp,
    showMinute: showMinuteProp,
    showHour: showHourProp,
    use12Hours: use12HoursProp
  } = props;
  const showSecond = showSecondProp !== null && showSecondProp !== void 0 ? showSecondProp : format ? generateFormat.showSecond(format) : false;
  const showMinute = showMinuteProp !== null && showMinuteProp !== void 0 ? showMinuteProp : format ? generateFormat.showMinute(format) : true;
  const showHour = showHourProp !== null && showHourProp !== void 0 ? showHourProp : format ? generateFormat.showHour(format) : true;
  const use12Hours = use12HoursProp !== null && use12HoursProp !== void 0 ? use12HoursProp : format ? generateFormat.is12Hour(format) : false;
  const bodyOperationRef = React.useRef();

  // ======================= Keyboard =======================
  const [activeColumnIndex, setActiveColumnIndex] = React.useState(-1);
  const columnsCount = countBoolean([showSecond, use12Hours]);
  operationRef && (operationRef.current = {
    onKeyDown: event => createKeyDownHandler(event, {
      onLeftRight: diff => {
        setActiveColumnIndex((activeColumnIndex + diff + columnsCount) % columnsCount);
      },
      onUpDown: diff => {
        if (activeColumnIndex === -1) {
          setActiveColumnIndex(0);
        } else if (bodyOperationRef.current) {
          bodyOperationRef.current.onUpDown(diff);
        }
      },
      onEnter: () => {
        onPanelChange(value || generateConfig.getNow());
        setActiveColumnIndex(-1);
      }
    }),
    onBlur: () => {
      setActiveColumnIndex(-1);
    }
  });
  return /*#__PURE__*/React.createElement("div", null, (() => {
    if (mode === 'timeCombine') {
      return /*#__PURE__*/React.createElement(CombineTimePanel, _extends({}, props, {
        use12Hours: use12Hours
      }));
    } else if (mode === 'time') {
      return /*#__PURE__*/React.createElement("div", {
        className: `${prefixCls}-time-panel`
      }, /*#__PURE__*/React.createElement(NormalTimePanel, _extends({}, props, {
        showMinute: showMinute,
        showHour: showHour,
        use12Hours: use12Hours,
        showSecond: showSecond,
        activeColumnIndex: activeColumnIndex
      })));
    }
  })());
}
