// @universe-design/react/src/cascader/CascaderOverlay.tsx
import type { Context } from 'react';
import React, { useContext } from 'react';
import type { TreeCheckProps, TreeNodeConfig, TreeSharedProps, genDoublyTreeNodeMap } from '../tree';
import { isNonLeaf } from '../tree';
import type { SelectAction, SelectContextType } from '../select';
import { SelectActionType, SelectContext } from '../select';
import useCascade from './cascade/useCascade';
import { CascadeUI } from './cascade/cascadeUI';
import { CascadeContextProvider } from './cascade/CascadeContext';
import { SelectOverlayContainer } from '../select/SelectOverlayContainer';
import clsx from 'clsx';
import { useCascaderMenuCssScope } from './common';
import type { CascadeSharedProps } from './cascade/interface';
import type { Required } from 'utility-types';
import type { CascadeRef } from './interface';

export const CascaderOverlay: React.FC<
  Required<CascadeSharedProps, 'changeOnSelect'> &
    ReturnType<typeof genDoublyTreeNodeMap> & { cascadeRef: React.Ref<CascadeRef> }
> = props => {
  const cssScope = useCascaderMenuCssScope();
  const { inputValue, valueSet, set, multiple } = useContext(
    SelectContext as Context<SelectContextType<string, TreeNodeConfig>>,
  );
  const { doublyTreeNodeMap, doublyTreeNodeList, changeOnSelect, showCheckedStrategy, cascadeRef, ...restProps } =
    props;

  const onSelect = React.useCallback<NonNullable<TreeSharedProps['onSelect']>>(
    (keys, _action) => {
      const action: SelectAction = {
        type: _action.selected ? SelectActionType.Select : SelectActionType.DeSelect,
        payload: _action.key,
      };

      let closeOverlay = true;
      if (multiple) {
        closeOverlay = false;
      } else if (changeOnSelect && isNonLeaf(doublyTreeNodeMap[_action.key].__node__)) {
        closeOverlay = false;
      }
      set(keys, { closeOverlay, action });
    },
    [changeOnSelect, doublyTreeNodeMap, multiple, set],
  );

  const onCheck = React.useCallback<NonNullable<TreeCheckProps['onCheck']>>(
    (keys, _action) => {
      const action: SelectAction = {
        type: _action.checked ? SelectActionType.Checked : SelectActionType.UnChecked,
        payload: _action.key,
      };
      set(keys, { action });
    },
    [set],
  );

  const cascadeCtxValue = useCascade(
    {
      ...restProps,
      changeOnSelect,
      value: Array.from(valueSet),
      searchValue: inputValue,
      multiple,
      onSelect,
      onCheck,
      showCheckedStrategy,
    },
    {
      doublyTreeNodeMap,
      doublyTreeNodeList,
    },
    cascadeRef,
  );

  const { searchFlattenTree } = cascadeCtxValue;
  const rootDoublyTreeNodeList = doublyTreeNodeList.filter(n => !n.parent);

  return (
    <CascadeContextProvider value={cascadeCtxValue}>
      <SelectOverlayContainer
        {...{
          className: clsx(
            cssScope(),
            (inputValue || !rootDoublyTreeNodeList.length) && cssScope('-search'), // 这个样式主要是为了设置 width 为 auto
          ),
          optionsLength: rootDoublyTreeNodeList.length,
          filteredOptionsLength: searchFlattenTree.length,
          menu: <CascadeUI />,
        }}
      />
    </CascadeContextProvider>
  );
};
