// @universe-design/react/src/cascader/index.tsx
/* eslint-disable react-hooks/rules-of-hooks */
import type { PropsWithChildren, ReactElement, Ref, RefAttributes } from 'react';
import React, { useMemo, useRef } from 'react';
import { getCascaderLabel } from './common';
import type {
  InternalSelectProps,
  SingleInternalSelectProps,
  MultipleInternalSelectProps,
  RefSelectRef,
} from '../select/internal.interface';
import { selectGenerator } from '../select';
import type { CascadeRef, CascaderProps, MultipleCascaderProps, SingleCascaderProps } from './interface';
import type { TreeNodeConfig } from '../tree';
import { ShowCheckedStrategy, getTreeNodeConfigPath, useTreeCheck } from '../tree';
import { devWarning, genArray, isNonNullable, warning } from '../util';
import { CascaderOverlay } from './CascaderOverlay';
import { useCascaderData } from './cascade/useCascaderData';
import './style';
import { strategyMap } from './cascade/useCascade';
import { TreeCheckedProvider } from '../tree/hooks/TreeCheckedContext';
import { COMPONENT_VERSION_ID } from '../config-provider';
import type { CascaderRef } from './cascade/interface';

const RefSelect = selectGenerator<string, TreeNodeConfig>({
  selectorCssScope: 'select',
  dropdownCssScope: 'cascader',
  // defaultOptionLabelProp: option => option.title || option.value,
});

const CascaderComp = React.forwardRef(function Cascader(
  props: PropsWithChildren<CascaderProps>,
  ref: Ref<CascaderRef>,
) {
  const {
    options = [],
    filterOption,
    filterResultSorter,
    filterResultLimit,
    filterResultRenderer,
    checkable,
    listHeight,
    dropdownLevelWidth,
    expandTrigger,
    fullPathBackfillOption = false,
    optionLabelProp: optionLabelPropProp,
    loadedKeys,
    loadData,
    onLoad,
    showCheckedStrategy = ShowCheckedStrategy.SHOW_CHILD,
    fullHeight,
    virtual,
    ...baseSelectProps
  } = props;
  const mode = baseSelectProps.mode ?? 'single';

  const selectRef = useRef<RefSelectRef>(null);

  const cascadeRef = useRef<CascadeRef>(null);

  const { doublyTreeNodeList, doublyTreeNodeMap, internalOptions } = useCascaderData({ options });

  const treeCheckedContext = useTreeCheck(
    {
      doublyTreeNodeMap,
      defaultCheckedKeys: genArray(props.defaultValue),
      ...('value' in props ? { checkedKeys: genArray(props.value) } : {}),
      checkStrategy: strategyMap[showCheckedStrategy],
      checkStrictly: false,
    },
    { skipped: !checkable },
  );

  const { checkedSet } = treeCheckedContext;
  const checkedKeys = React.useMemo(() => Object.keys(checkedSet).filter(key => checkedSet[key]), [checkedSet]);
  if (mode === 'multiple' && checkable) {
    if ('value' in props) {
      // 受控模式
      baseSelectProps.value = checkedKeys;
    } else {
      // 非受控模式
      baseSelectProps.defaultValue = checkedKeys;
    }
  }

  /**
   * Select Props
   */
  const optionLabelProp = useMemo(
    () =>
      ({ value }: TreeNodeConfig) =>
        doublyTreeNodeMap[value]
          ? getCascaderLabel(doublyTreeNodeMap[value], {
              highlightMarkClassName: '',
              cascaded: fullPathBackfillOption,
              optionLabelProp: optionLabelPropProp,
            })
          : value,
    [doublyTreeNodeMap, fullPathBackfillOption, optionLabelPropProp],
  );

  baseSelectProps.mode = mode;
  const changeOnSelect = baseSelectProps.changeOnSelect ?? mode !== 'single';
  baseSelectProps.changeOnSelect = changeOnSelect;

  warning(!(mode === 'single' && checkable), devWarning('Cascader', 'Single check mode does not support.'));

  warning(
    !(mode === 'multiple' && !checkable && changeOnSelect),
    devWarning('Cascader', 'Multiple select mode changeOnSelect=true does not support.'),
  );

  warning(
    !(mode === 'multiple' && checkable && !changeOnSelect),
    devWarning('Cascader', 'Multiple check mode changeOnSelect=false does not support.'),
  );

  const overlayNode = (
    <TreeCheckedProvider value={treeCheckedContext}>
      <CascaderOverlay
        cascadeRef={cascadeRef}
        {...{
          loadData,
          loadedKeys,
          onLoad,
          dropdownLevelWidth,
          listHeight,
          filterOption,
          filterResultSorter,
          filterResultLimit,
          filterResultRenderer,
          expandTrigger,
          changeOnSelect,
          checkable,
          doublyTreeNodeList,
          doublyTreeNodeMap,
          showCheckedStrategy,
          fullHeight,
          virtual,
        }}
      />
    </TreeCheckedProvider>
  );

  React.useImperativeHandle(
    ref,
    () => ({
      blur() {
        selectRef.current?.blur();
      },
      focus() {
        selectRef.current?.focus();
      },
      get input() {
        return selectRef.current?.input;
      },
      scrollIntoView(...args) {
        return cascadeRef.current?.scrollIntoView(...args);
      },
    }),
    [cascadeRef, selectRef],
  );

  let selectProps: InternalSelectProps<string, TreeNodeConfig>;

  if (baseSelectProps.mode === 'multiple') {
    const _onChange: MultipleInternalSelectProps<string, TreeNodeConfig>['onChange'] = (value, meta) => {
      if (baseSelectProps.onChange) {
        const path = value.length
          ? value.reduce((a, b, index) => {
              const _path = getTreeNodeConfigPath(doublyTreeNodeMap[b]);
              a[index] = _path;
              return a;
            }, [] as TreeNodeConfig[][])
          : [];
        baseSelectProps.onChange(value, { ...meta, path });
      }
    };

    selectProps = {
      ...baseSelectProps,
      mode: 'multiple',
      options: internalOptions,
      overlayNode,
      onChange: _onChange,
    };
  } else {
    const _onChange: SingleInternalSelectProps<string, TreeNodeConfig>['onChange'] = (value, meta) => {
      if (baseSelectProps.onChange) {
        const path = getTreeNodeConfigPath(isNonNullable(value) ? doublyTreeNodeMap[value] : null);
        baseSelectProps.onChange(value, { ...meta, path });
      }
    };

    selectProps = {
      ...baseSelectProps,
      mode: 'single',
      options: internalOptions,
      overlayNode,
      onChange: _onChange,
    };
  }

  return <RefSelect<string, TreeNodeConfig> {...selectProps} optionLabelProp={optionLabelProp} ref={selectRef} />;
}) as {
  (
    props: (PropsWithChildren<MultipleCascaderProps> | PropsWithChildren<SingleCascaderProps>) &
      RefAttributes<CascaderRef>,
  ): ReactElement | null;
  displayName: string;
};

export const Cascader = Object.assign(CascaderComp, {
  SHOW_PARENT: ShowCheckedStrategy.SHOW_PARENT,
  SHOW_CHILD: ShowCheckedStrategy.SHOW_CHILD,
  [COMPONENT_VERSION_ID]: process.env.__VERSION__,
} as const);

if (__DEV__) {
  Cascader.displayName = 'Cascader';
}

export type { CascaderProps, MultipleCascaderProps, SingleCascaderProps, CascaderSelectorProps } from './interface';

export type {
  CascadeSharedProps,
  CascaderFilterOptionFunc,
  CascaderRef,
  CascaderVirtualParams,
} from './cascade/interface';
export type { defaultFilterOption } from './common';
