// @universe-design/react/src/cascader/interface.ts
import type React from 'react';
import type {
  MultipleInternalSelectProps as _MultipleInternalSelectProps,
  SingleInternalSelectProps as _SingleInternalSelectProps,
  BaseInternalSelectProps,
} from '../select/internal.interface';
import type { TreeNodeConfig } from '../tree';
import type { CascadeSharedProps } from './cascade/interface';
import type { ListScrollConfig } from '..';

/**
 * @public
 */
export interface SingleInternalSelectProps extends _SingleInternalSelectProps<string, TreeNodeConfig> {
  checkable?: undefined | false;
}

interface MultipleInternalSelectProps extends _MultipleInternalSelectProps<string, TreeNodeConfig> {
  checkable?: boolean;
}
/**
 * @public
 */
export type CascaderSelectorProps = Omit<
  BaseInternalSelectProps<string, TreeNodeConfig>,
  'optionLabelProp' | 'dropdownMatchSelectWidth'
> & {
  /**
   * 回填到选择框的选项是否级联
   * @defaultValue false
   */
  fullPathBackfillOption?: boolean;
  /**
   * 回填到选择框的选项属性值，默认是选项的子元素
   */
  optionLabelProp?: string | ((node: TreeNodeConfig, meta: { path: TreeNodeConfig[] }) => React.ReactNode);
  /**
   * Tree 型数据结构，嵌套数组构造的树
   */
  options?: TreeNodeConfig[];
  /**
   * 搜索结果面板是否和选择器输入框宽度相同，如果参数为 `number` 类型则作为下拉菜单的宽度。
   * @defaultValue false
   */
  dropdownMatchSelectWidth?: boolean | number;
};

type OriginSingleOnChangeParams = Parameters<NonNullable<SingleInternalSelectProps['onChange']>>;
type OriginMultipleOnChangeParams = Parameters<NonNullable<MultipleInternalSelectProps['onChange']>>;

export type SingleOnChangeType = (
  value: OriginSingleOnChangeParams[0],
  meta: OriginSingleOnChangeParams[1] & { path: TreeNodeConfig[] },
) => void;

export type MultipleOnChangeType = (
  value: OriginMultipleOnChangeParams[0],
  meta: OriginMultipleOnChangeParams[1] & {
    path: TreeNodeConfig[][];
  },
) => void;

/**
 * @internal
 */
export type SingleCascaderProps = (Omit<SingleInternalSelectProps, 'onChange'> & { onChange?: SingleOnChangeType }) &
  CascaderSelectorProps &
  Omit<Partial<CascadeSharedProps>, 'showCheckedStrategy'>;

/**
 * @internal
 */
export type MultipleCascaderProps = (Omit<MultipleInternalSelectProps, 'onChange'> & {
  onChange?: MultipleOnChangeType;
}) &
  CascaderSelectorProps &
  Partial<CascadeSharedProps>;

/**
 * @public
 * @docgen
 */
export type CascaderProps = (
  | (Omit<SingleInternalSelectProps, 'onChange'> & {
      onChange?: SingleOnChangeType;
    })
  | (Omit<MultipleInternalSelectProps, 'onChange'> & {
      onChange?: MultipleOnChangeType;
    })
) &
  CascaderSelectorProps &
  Partial<CascadeSharedProps>;

export interface CascaderRef {
  /**
   * 选择器聚焦方法
   *
   * @public
   */
  focus: HTMLOrSVGElement['focus'];
  /**
   * 选择器失焦方法
   *
   * @public
   */
  blur: HTMLOrSVGElement['blur'];
  /**
   * 选择器输入框对应的 DOM 元素，是一个 getter
   *
   * @public
   */
  readonly input: HTMLInputElement | undefined | null;
  /**
   * 将指定选项滚动到可视区域内
   *
   * @public
   */
  scrollIntoView: (key: string, scrollConfig?: ListScrollConfig) => void;
}

export interface CascadeRef {
  /**
   * 将指定选项滚动到可视区域内
   *
   * @public
   */
  scrollIntoView: (key: string, scrollConfig?: ListScrollConfig) => void;
}
