// @universe-design/react/src/radio/stories/homeless.stories.tsx
import { Radio, Space, Switch } from '../..';
import React, { useState } from 'react';

export default {
  title: 'Components/Radio/临时示例',
  component: Radio,
};

export const SupportBooleanValueInGroup = () => (
  <>
    <Radio.Group onChange={v => console.log(typeof v, v)}>
      <Radio value={true}>1</Radio>
      <Radio value={false}>2</Radio>
    </Radio.Group>
  </>
);

export const SupportBooleanValue = () => (
  <>
    <Radio value={true} onChange={v => console.log(typeof (v.target as any).value)}>
      1
    </Radio>
  </>
);

export const EmptyValue = () => (
  <>
    <Radio.Group onChange={v => console.log(typeof v, v)}>
      <Radio>1</Radio>
      <Radio>2</Radio>
    </Radio.Group>
  </>
);

export const StringOption = () => (
  <>
    <Radio.Group options={['a', 'b', 'c']} onChange={v => console.log(typeof v, v)} />
  </>
);

export const Error = () => {
  const [error, setError] = useState<boolean>(false);
  const [disabled, setDisabled] = useState<boolean>(false);
  return (
    <Space style={{ width: '100%' }} direction="vertical" size="large">
      <Space align="center">
        <span>错误：</span>
        <Switch checked={error} onChange={setError} />
      </Space>
      <Space align="center">
        <span>禁用：</span>
        <Switch checked={disabled} onChange={setDisabled} />
      </Space>
      <Radio error={error} disabled={disabled} />
      <Radio.Group error={error} disabled={disabled}>
        <Radio value={0}>Apple</Radio>
        <Radio value={1}>Pear</Radio>
        <Radio value={2}>Orange</Radio>
        <Radio value={3}>Orange</Radio>
      </Radio.Group>
      <Radio.Group error={error} disabled={disabled}>
        <Radio.Button value={0}>Apple</Radio.Button>
        <Radio.Button value={1}>Pear</Radio.Button>
        <Radio.Button value={2}>Orange</Radio.Button>
        <Radio.Button value={3}>Orange</Radio.Button>
      </Radio.Group>
      <Radio.Group buttonStyle="solid" error={error} disabled={disabled}>
        <Radio.Button value={0}>Apple</Radio.Button>
        <Radio.Button value={1}>Pear</Radio.Button>
        <Radio.Button value={2}>Orange</Radio.Button>
        <Radio.Button value={3}>Orange</Radio.Button>
      </Radio.Group>
    </Space>
  );
};

export { default as OptionTypeAll } from '../demo/option-type-all';
