// @universe-design/react/src/radio/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C8VerticalAlign from '../demo/8.vertical-align';
import CBasic from '../demo/basic';
import CButtonStyle from '../demo/button-style';
import CButton from '../demo/button';
import CDisabled from '../demo/disabled';
import CGroupDirection from '../demo/group.direction';
import CGroup from '../demo/group';
import COption from '../demo/option';
import CSize from '../demo/size';
if (!(C8VerticalAlign as Story).storyName) {
  (C8VerticalAlign as Story).storyName = '7. 垂直对齐';
  (C8VerticalAlign as Story).parameters = {
    docs: {
      description: {
        story: '单选文本内容有两种对齐方式，默认采用首行垂直对齐 `align-items: baseline;`。\n\n',
      },
    },
  };
}
if (!(CBasic as Story).storyName) {
  (CBasic as Story).storyName = '0. 基础用法';
  (CBasic as Story).parameters = {
    docs: {
      description: {
        story: '简单的用法。\n\n',
      },
    },
  };
}
if (!(CButtonStyle as Story).storyName) {
  (CButtonStyle as Story).storyName = '5. 按钮样式 - 填充风格';
  (CButtonStyle as Story).parameters = {
    docs: {
      description: {
        story: '填底的单选按钮样式。\n\n',
      },
    },
  };
}
if (!(CButton as Story).storyName) {
  (CButton as Story).storyName = '4. 按钮样式';
  (CButton as Story).parameters = {
    docs: {
      description: {
        story: '按钮样式的单选组合。\n\n',
      },
    },
  };
}
if (!(CDisabled as Story).storyName) {
  (CDisabled as Story).storyName = '1. 禁用状态';
  (CDisabled as Story).parameters = {
    docs: {
      description: {
        story: '单选器不可用。\n\n',
      },
    },
  };
}
if (!(CGroupDirection as Story).storyName) {
  (CGroupDirection as Story).storyName = '3. 单选组 - 垂直';
  (CGroupDirection as Story).parameters = {
    docs: {
      description: {
        story: '设置 `direction` 为 `vertical` 可以展示竖直的单选组。\n\n',
      },
    },
  };
}
if (!(CGroup as Story).storyName) {
  (CGroup as Story).storyName = '2. 单选组';
  (CGroup as Story).parameters = {
    docs: {
      description: {
        story: '一组互斥的单选器配合使用。\n\n',
      },
    },
  };
}
if (!(COption as Story).storyName) {
  (COption as Story).storyName = '7. 单选组 - 配置方式';
  (COption as Story).parameters = {
    docs: {
      description: {
        story: '通过配置 `options` 参数来渲染单选框。也可通过 `optionType` 参数来设置单选器类型。\n\n',
      },
    },
  };
}
if (!(CSize as Story).storyName) {
  (CSize as Story).storyName = '6. 按钮样式 - 大小';
  (CSize as Story).parameters = {
    docs: {
      description: {
        story: '通过设置单选组的 `size` 属性，来改变按钮的大小。可以和表单输入框进行对应配合。只在按钮类型下生效。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Radio/官网示例',
  tags: ['autodocs'],
};

export { C8VerticalAlign, CBasic, CButtonStyle, CButton, CDisabled, CGroupDirection, CGroup, COption, CSize };
