// @universe-design/react/src/rate/index.tsx
import type { FC, ReactNode, CSSProperties, KeyboardEventHandler } from 'react';
import React, { useState, useRef, useEffect } from 'react';
import clsx from 'clsx';
import stylex from '@universe-design/stylex';
import { useControlledState } from '@universe-design/react-utilities';
import { cssTypographyBody0 } from '../stylex';
import { useCssScope, useLocale } from '../config-provider/hooks';
import { Tooltip } from '../tooltip';
import CollectFilled from '@universe-design/icons-react/CollectFilled';
import { isFunction } from '../util';
import noop from 'lodash/noop';

import './style';

export interface RateProps {
  /**
   * `className` 属性透传
   */
  className?: string;
  /**
   * `style` 属性透传
   */
  style?: CSSProperties;
  /**
   * 评分值受控属性
   */
  value?: number;
  /**
   * 默认评分值
   * @defaultValue 0
   */
  defaultValue?: number;
  /**
   * 评分星星总数
   * @defaultValue 5
   */
  count?: number;
  /**
   * 是否允许半星选择
   * @defaultValue false
   */
  allowHalf?: boolean;
  /**
   * 是否允许再次点击后清除
   * @defaultValue true
   */
  allowClear?: boolean;
  /**
   * 自定义字符
   * @defaultValue `<CollectFilled />`
   */
  character?: ReactNode | ((props: RateProps & { index: number }) => ReactNode);
  /**
   * 自动获取焦点
   * @defaultValue false
   */
  autoFocus?: boolean;
  /**
   * @deprecated 看代码里面没有用这个
   */
  direction?: string;
  /**
   * 是否只读
   * @defaultValue false
   */
  readonly?: boolean;
  /**
   * 自定义每项的提示信息
   */
  tooltips?: string[];
  /**
   * 是否显示描述信息
   * @defaultValue true
   */
  showDescriptions?: boolean;
  /**
   * 自定义每项的描述信息，显示在评分组件下方
   */
  descriptions?: string[];
  /**
   * 选择时的回调
   */
  onChange?: (value: number) => void;
  /**
   * 鼠标经过时数值变化的回调
   */
  onHoverChange?: (value: number) => void;
  /**
   * 获取焦点时的回调
   */
  onFocus?: () => void;
  /**
   * 失去焦点时的回调
   */
  onBlur?: () => void;
  /**
   * 按键回调
   */
  onKeyDown?: KeyboardEventHandler<HTMLUListElement>;
}

export const Rate: FC<RateProps> = props => {
  const {
    style,
    className,
    character,
    allowHalf,
    allowClear = true,
    value: valueInProps,
    defaultValue = 0,
    readonly = false,
    count = 5,
    tooltips = [],
    showDescriptions = true,
    descriptions = [],
    autoFocus = false,
    onHoverChange,
    onChange,
  } = props;

  const cssScope = useCssScope('rate');
  const cls = clsx(cssScope(), className);
  const {
    components: { rate: rateLocale },
  } = useLocale();

  const rateRef = useRef<HTMLDivElement>(null);
  const [value, setValue] = useControlledState(defaultValue, valueInProps, {
    onChange,
  });
  const [hoverValue, setHoverValue] = useState<number>();
  const [cleanedValue, setCleanedValue] = useState<number>();

  useEffect(() => {
    if (rateRef.current && autoFocus) {
      rateRef.current.focus();
    }
  }, [rateRef, autoFocus]);

  const getValue = (index: number, e: React.MouseEvent<HTMLDivElement, MouseEvent>): number => {
    let currentValue = index + 1;
    const { left, width } = e.currentTarget.getBoundingClientRect();
    if (allowHalf && e.clientX - left < width / 2) {
      // eslint-disable-next-line @typescript-eslint/no-magic-numbers
      currentValue -= 0.5;
    }
    return currentValue;
  };

  const calculateDes = () => {
    const currentValue = hoverValue || value || 0;
    const index = parseInt(currentValue.toFixed(), 10);
    return descriptions[index] || rateLocale.descriptions[index];
  };

  const handleMouseMove = (index: number, e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
    const currentValue = getValue(index, e);
    // 避免点击触发清除后，再次设置选中态
    /* istanbul ignore else  */
    if (currentValue !== cleanedValue) {
      setHoverValue(currentValue);
      isFunction(onHoverChange) && onHoverChange(currentValue);
    }
  };

  const handleMouseLeave = () => {
    setHoverValue(undefined);
    setCleanedValue(undefined);
    isFunction(onHoverChange) && onHoverChange(0);
  };

  const handleChoose = (index: number, e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
    const currentValue = getValue(index, e);
    const needReset = allowClear && currentValue === value;
    setValue(needReset ? 0 : currentValue);
    setHoverValue(undefined);
    setCleanedValue(needReset ? currentValue : undefined);
  };

  const wrapTooltips = (index: number, node: ReactNode) => {
    const tip = tooltips[index];
    if (tip) {
      return (
        <Tooltip key={`rate-tooltip-${index}`} title={tip}>
          {node}
        </Tooltip>
      );
    }

    return node;
  };

  const renderCharacterEl = (index: number) => {
    if (isFunction(character)) {
      return character({ ...props, index });
    } else {
      return character || <CollectFilled />;
    }
  };

  const renderCharacter = (index: number) => {
    let preciseStyle;
    const currentIndex = index + 1;
    const currentValue = hoverValue || value;
    const includes = currentValue && currentValue > index && currentValue < currentIndex;
    const hasHalf = allowHalf ? includes : false;
    const itemCls = clsx(cssScope('__character'), {
      [cssScope('__character--full')]: currentValue && currentValue >= currentIndex,
      [cssScope('__character--half')]: (readonly && includes) || hasHalf,
    });

    if (readonly && currentValue && includes) {
      preciseStyle = {
        width: `${(currentValue - index) * 100}%`,
      };
    }

    return wrapTooltips(
      index,
      <div
        className={itemCls}
        key={index}
        onMouseMove={readonly ? noop : e => handleMouseMove(index, e)}
        onClick={readonly ? noop : e => handleChoose(index, e)}
        data-testid={cssScope(`__character-${index}`)}
      >
        <div style={preciseStyle} className={cssScope('__character-half-body')}>
          <div className={cssScope('__character-half-body-wrap')}>{renderCharacterEl(index)}</div>
        </div>
        <div className={cssScope('__character-body')}>{renderCharacterEl(index)}</div>
      </div>,
    );
  };

  const renderCharacters = () => {
    const items: ReactNode[] = [];
    for (let i = 0; i < count; i++) {
      items.push(renderCharacter(i));
    }
    return items;
  };

  const renderRateDes = () => {
    if (showDescriptions) {
      return <div className={clsx(cssScope('__des'), stylex(cssTypographyBody0))}>{calculateDes()}</div>;
    }
    return null;
  };

  return (
    <div ref={rateRef} style={style} className={cls} onMouseLeave={handleMouseLeave}>
      <div className={cssScope('__characters')}>{renderCharacters()}</div>
      {renderRateDes()}
    </div>
  );
};
