// @universe-design/react/src/shared/components/picker/utils/ui.ts
import { isFunction } from '../../../../util';
import type { GenerateConfig } from '../generate';
import type { IFormatOption, IParseOption, Locale, MODE } from '../interface';

function getDefaultFormat(mode?: MODE, locale?: Locale) {
  switch (mode) {
    case 'year':
      return 'YYYY';
    case 'month':
      return 'YYYY-MM';
    case 'time':
      return 'HH:mm';
    case 'timeCombine':
      return 'HH:mm';
    case 'date':
    default:
      return 'YYYY-MM-DD';
  }
}

export function formatValue<DateType>(
  generateConfig: GenerateConfig<DateType>,
  value?: DateType,
  option?: IFormatOption<DateType>,
) {
  if (!value || !option) {
    return '';
  }

  const { mode, locale, localeName } = option;

  let { format } = option;

  if (!format) {
    format = getDefaultFormat(mode, locale);
  }

  if (isFunction(format)) {
    return format(value);
  }

  return generateConfig.locale.format(localeName, value, format);
}

export function parseValue<DateType>(
  generateConfig: GenerateConfig<DateType>,
  value?: string,
  option?: IParseOption<DateType>,
) {
  if (!value || !option) {
    return null;
  }

  const { mode, locale, localeName, parse } = option;
  let { format } = option;

  if (isFunction(parse)) {
    return parse(value);
  }

  if (!format || typeof format !== 'string') {
    format = getDefaultFormat(mode, locale);
  }

  return generateConfig.locale.parse(localeName, value, format);
}

export function elementsContains(elements: (HTMLElement | undefined | null)[], target: HTMLElement) {
  return elements.some(ele => ele && ele.contains(target));
}
