// @universe-design/styles/cjs/ccm-ccm-chat/index.light.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ccmtoken_chart_aurora_y200 = exports.ccmtoken_chart_aurora_y100 = exports.ccmtoken_chart_aurora_w600 = exports.ccmtoken_chart_aurora_w500 = exports.ccmtoken_chart_aurora_w400 = exports.ccmtoken_chart_aurora_w300 = exports.ccmtoken_chart_aurora_w200 = exports.ccmtoken_chart_aurora_w100 = exports.ccmtoken_chart_aurora_t600 = exports.ccmtoken_chart_aurora_t500 = exports.ccmtoken_chart_aurora_t400 = exports.ccmtoken_chart_aurora_t300 = exports.ccmtoken_chart_aurora_t200 = exports.ccmtoken_chart_aurora_t100 = exports.ccmtoken_chart_aurora_r600 = exports.ccmtoken_chart_aurora_r500 = exports.ccmtoken_chart_aurora_r400 = exports.ccmtoken_chart_aurora_r300 = exports.ccmtoken_chart_aurora_r200 = exports.ccmtoken_chart_aurora_r100 = exports.ccmtoken_chart_aurora_p600 = exports.ccmtoken_chart_aurora_p500 = exports.ccmtoken_chart_aurora_p400 = exports.ccmtoken_chart_aurora_p300 = exports.ccmtoken_chart_aurora_p200 = exports.ccmtoken_chart_aurora_p100 = exports.ccmtoken_chart_aurora_o600 = exports.ccmtoken_chart_aurora_o500 = exports.ccmtoken_chart_aurora_o400 = exports.ccmtoken_chart_aurora_o300 = exports.ccmtoken_chart_aurora_o200 = exports.ccmtoken_chart_aurora_o100 = exports.ccmtoken_chart_aurora_g600 = exports.ccmtoken_chart_aurora_g500 = exports.ccmtoken_chart_aurora_g400 = exports.ccmtoken_chart_aurora_g300 = exports.ccmtoken_chart_aurora_g200 = exports.ccmtoken_chart_aurora_g100 = exports.ccmtoken_chart_aurora_c600 = exports.ccmtoken_chart_aurora_c500 = exports.ccmtoken_chart_aurora_c400 = exports.ccmtoken_chart_aurora_c300 = exports.ccmtoken_chart_aurora_c200 = exports.ccmtoken_chart_aurora_c100 = exports.ccmtoken_chart_aurora_b600 = exports.ccmtoken_chart_aurora_b500 = exports.ccmtoken_chart_aurora_b400 = exports.ccmtoken_chart_aurora_b300 = exports.ccmtoken_chart_aurora_b200 = exports.ccmtoken_chart_aurora_b100 = void 0;
exports.ccmtoken_chart_aurora_y600 = exports.ccmtoken_chart_aurora_y500 = exports.ccmtoken_chart_aurora_y400 = exports.ccmtoken_chart_aurora_y300 = void 0;
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
exports.ccmtoken_chart_aurora_b100 = '#bdd2ff';
exports.ccmtoken_chart_aurora_b200 = '#8fb4ff';
exports.ccmtoken_chart_aurora_b300 = '#4c88ff';
exports.ccmtoken_chart_aurora_b400 = '#3370eb';
exports.ccmtoken_chart_aurora_b500 = '#2655b6';
exports.ccmtoken_chart_aurora_b600 = '#194294';
exports.ccmtoken_chart_aurora_c100 = '#ffc2e5';
exports.ccmtoken_chart_aurora_c200 = '#fc94cf';
exports.ccmtoken_chart_aurora_c300 = '#ed77ba';
exports.ccmtoken_chart_aurora_c400 = '#c24a8e';
exports.ccmtoken_chart_aurora_c500 = '#94386c';
exports.ccmtoken_chart_aurora_c600 = '#782b57';
exports.ccmtoken_chart_aurora_g100 = '#99e490';
exports.ccmtoken_chart_aurora_g200 = '#69cc5c';
exports.ccmtoken_chart_aurora_g300 = '#51ba43';
exports.ccmtoken_chart_aurora_g400 = '#419e34';
exports.ccmtoken_chart_aurora_g500 = '#2f7526';
exports.ccmtoken_chart_aurora_g600 = '#21511a';
exports.ccmtoken_chart_aurora_o100 = '#fec88b';
exports.ccmtoken_chart_aurora_o200 = '#f89e44';
exports.ccmtoken_chart_aurora_o300 = '#f3871b';
exports.ccmtoken_chart_aurora_o400 = '#db7018';
exports.ccmtoken_chart_aurora_o500 = '#a15317';
exports.ccmtoken_chart_aurora_o600 = '#683a12';
exports.ccmtoken_chart_aurora_p100 = '#dbc8fd';
exports.ccmtoken_chart_aurora_p200 = '#c5a3ff';
exports.ccmtoken_chart_aurora_p300 = '#a575fa';
exports.ccmtoken_chart_aurora_p400 = '#8c55ec';
exports.ccmtoken_chart_aurora_p500 = '#6c39c6';
exports.ccmtoken_chart_aurora_p600 = '#5529a3';
exports.ccmtoken_chart_aurora_r100 = '#f89e9b';
exports.ccmtoken_chart_aurora_r200 = '#f6827e';
exports.ccmtoken_chart_aurora_r300 = '#f05b56';
exports.ccmtoken_chart_aurora_r400 = '#d14642';
exports.ccmtoken_chart_aurora_r500 = '#a03331';
exports.ccmtoken_chart_aurora_r600 = '#a03331';
exports.ccmtoken_chart_aurora_t100 = '#6fe8d8';
exports.ccmtoken_chart_aurora_t200 = '#33d6c0';
exports.ccmtoken_chart_aurora_t300 = '#2dbeab';
exports.ccmtoken_chart_aurora_t400 = '#10a893';
exports.ccmtoken_chart_aurora_t500 = '#0f8575';
exports.ccmtoken_chart_aurora_t600 = '#1d4e47';
exports.ccmtoken_chart_aurora_w100 = '#89dffe';
exports.ccmtoken_chart_aurora_w200 = '#51c3f0';
exports.ccmtoken_chart_aurora_w300 = '#25b2e5';
exports.ccmtoken_chart_aurora_w400 = '#1099cc';
exports.ccmtoken_chart_aurora_w500 = '#146c91';
exports.ccmtoken_chart_aurora_w600 = '#164359';
exports.ccmtoken_chart_aurora_y100 = '#ffde75';
exports.ccmtoken_chart_aurora_y200 = '#fbcb46';
exports.ccmtoken_chart_aurora_y300 = '#f0b622';
exports.ccmtoken_chart_aurora_y400 = '#d99904';
exports.ccmtoken_chart_aurora_y500 = '#ad7a03';
exports.ccmtoken_chart_aurora_y600 = '#6f4a01';
