// @universe-design/react/src/radio/group.tsx
import type { CSSProperties, ReactNode } from 'react';
import React, { forwardRef } from 'react';
import clsx from 'clsx';
import { useControlledState } from '@universe-design/react-utilities';
import Radio from './radio';
import { useComponentSize, useCssScope } from '../config-provider/hooks';
import { RadioGroupContextProvider } from './context';
import type { DefaultSizeType } from '../util';
import { isFunction, getDefaultSizeAbbr } from '../util';
import type { RadioChangeEvent, RadioGroupOptionType, RadioValue } from './interface';

import './style';
/**
 * @public
 */
export type RadioGroupOptionMode = 'outline' | 'solid';
/**
 * @public
 */
export type RadioGroupDirection = 'horizontal' | 'vertical';

export interface RadioGroupOption {
  /**
   * `stlye` 属性透传
   */
  style?: CSSProperties;
  /**
   * `className` 属性透传
   */
  className?: string;
  /**
   * 单选器标签
   */
  label: ReactNode;
  /**
   * 单选器的值
   */
  value: RadioValue;
  /**
   * 是否禁用
   * @defaultValue false
   */
  disabled?: boolean;
  /**
   * 错误状态
   * @defaultValue false
   */
  error?: boolean;
}

export type OptionsType = RadioGroupOption | string;

/**
 * 单选框组合，用于包裹一组单选器
 */
export interface RadioGroupProps {
  /**
   * 自定义类名前缀
   */
  prefixCls?: string;
  /**
   * 自定义 id 属性
   */
  id?: string;
  /**
   * 大小，只对按钮样式生效
   */
  size?: DefaultSizeType;
  /**
   * 单选组下所有 `input[type="radio"]` 的 `name` 属性
   */
  name?: string;
  /**
   * 自定义样式
   */
  style?: CSSProperties;
  /**
   * 自定义类名
   */
  className?: string;
  /**
   * 以配置形式设置子元素
   */
  options?: OptionsType[];
  /**
   * 禁选所有子单选器
   */
  disabled?: boolean;
  /**
   * 错误状态
   * @defaultValue false
   */
  error?: boolean;
  /**
   * 用于设置当前选中的值
   */
  value?: RadioValue;
  /**
   * 默认选中的值
   */
  defaultValue?: RadioValue;
  /**
   * 用于设置单选器的 `options` 类型
   * @defaultValue `default`
   */
  optionType?: RadioGroupOptionType;
  /**
   * 单选按钮的风格样式，`outline` 描边风格已废弃
   * @defaultValue `outline`
   */
  buttonStyle?: RadioGroupOptionMode;
  /**
   * 方向
   * @defaultValue `horizontal`
   */
  direction?: RadioGroupDirection;
  /**
   * 自定义子节点
   */
  children?: React.ReactNode;
  /**
   * 选项变化时的回调函数
   */
  onChange?: (value: RadioValue, e: RadioChangeEvent) => void;
  /**
   * 鼠标移入事件
   */
  onMouseEnter?: React.MouseEventHandler<HTMLDivElement>;
  /**
   * 鼠标移出事件
   */
  onMouseLeave?: React.MouseEventHandler<HTMLDivElement>;
}

/**
 * 如果有 radio button child，父级 optionType 就要是 button
 * https://stackoverflow.com/questions/21252551/apply-style-to-parent-if-it-has-child-with-css
 */
function useRadioGroupButtonOptionType({
  optionType: optionTypeProp = 'default',
}: {
  optionType?: RadioGroupOptionType;
}) {
  const [optionTypeRegistry, setButtonRegistry] = React.useState(
    new Map<string | number | boolean, RadioGroupOptionType>(),
  );
  const registerBOT = React.useCallback((value: string | number | boolean) => {
    setButtonRegistry(s => new Map([...s.entries(), [value, 'button']]));
  }, []);
  const unregisterBOT = React.useCallback((value: string | number | boolean) => {
    setButtonRegistry(s => {
      s.delete(value);
      return new Map([...s.entries()]);
    });
  }, []);
  const optionType = optionTypeRegistry.size ? 'button' : optionTypeProp;
  return {
    registerBOT,
    unregisterBOT,
    optionType,
  };
}

const RadioGroup = forwardRef<HTMLDivElement, RadioGroupProps>((props, ref) => {
  const {
    className,
    options,
    direction = 'horizontal',
    // buttonStyle = 'outline',
    disabled,
    children,
    size,
    style,
    id,
    onMouseEnter,
    onMouseLeave,
    prefixCls,
  } = props;
  const ctxComponentSize = useComponentSize();
  const mergedSize = size ?? ctxComponentSize;
  const groupCssScope = useCssScope('radio-group', prefixCls);
  const radioButtonScope = useCssScope('radio-button', prefixCls)();
  const [value, setValue] = useControlledState(props.defaultValue, props.value, {
    isControlledOverride: 'value' in props,
  });

  const onRadioChange = (ev: RadioChangeEvent) => {
    const lastValue = value;
    const val = ev.target.value;
    // 在 group 模式下，如果没有提供 val，不会触发 onChange。
    if (val === undefined) {
      return;
    }
    setValue(val);
    const { onChange } = props;
    isFunction(onChange) && val !== lastValue && onChange(val, ev);
  };

  const { registerBOT, unregisterBOT, optionType } = useRadioGroupButtonOptionType(props);

  const renderGroup = () => {
    let renderRadio = children;
    if (options && options.length > 0) {
      const radioPrefixCls = optionType === 'button' ? radioButtonScope : undefined;
      renderRadio = options.map(option => {
        if (typeof option === 'string') {
          return (
            <Radio
              key={option}
              disabled={disabled}
              value={option}
              checked={value === option}
              prefixCls={radioPrefixCls}
              optionType={optionType}
            >
              {option}
            </Radio>
          );
        }
        return (
          <Radio
            key={`radio-option-${option.value}`}
            disabled={option.disabled || disabled}
            value={option.value}
            checked={value === option.value}
            style={option.style}
            className={option.className}
            prefixCls={radioPrefixCls}
            optionType={optionType}
          >
            {option.label}
          </Radio>
        );
      });
    }

    const componentSize = getDefaultSizeAbbr(mergedSize);
    const classNames = clsx(
      groupCssScope(),
      optionType === 'button' && groupCssScope('-button'),
      groupCssScope('--mode-solid'), // 目前我们的设计只有 solid mode 一种样式
      groupCssScope(`--direction-${direction}`),
      {
        [groupCssScope(`--size-${componentSize}`)]: componentSize,
      },
      className,
    );
    return (
      <div
        className={classNames}
        style={style}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
        id={id}
        ref={ref}
      >
        {renderRadio}
      </div>
    );
  };

  return (
    <RadioGroupContextProvider
      value={{
        onChange: onRadioChange,
        value,
        disabled: props.disabled,
        name: props.name,
        error: props.error,
        registerBOT,
        unregisterBOT,
        size: mergedSize,
      }}
    >
      {renderGroup()}
    </RadioGroupContextProvider>
  );
});

export default React.memo(RadioGroup);
