// @universe-design/react/src/radio/radio.tsx
import type { CSSProperties, MouseEventHandler, ChangeEvent } from 'react';
import React, { useRef, forwardRef, useImperativeHandle } from 'react';
import clsx from 'clsx';
import type { StylexInstance } from '@universe-design/stylex';
import stylex from '@universe-design/stylex';
import { useControlledState } from '@universe-design/react-utilities';
import { cssFontMedium, cssTypographyBody2, cssTypographyBody0, cssBlockBase } from '../stylex';
import useRadioGroupContext from './useRadioGroupContext';
import { useCssScope } from '../config-provider/hooks';
import { isFunction, omit } from '../util';
import type { RadioChangeEvent, RadioGroupOptionType, RadioValue } from './interface';
import { Text } from '../text';
import './style';

const RADIO_LABEL_MARGIN = 8;
const RADIO_BUTTON_MAX_WIDTH = 240;

type HTMLInputPropsType = React.InputHTMLAttributes<HTMLInputElement>;
export interface RadioProps extends Omit<HTMLInputPropsType, 'onChange' | 'value' | 'onMouseEnter' | 'onMouseLeave'> {
  prefixCls?: string;
  /**
   * `className`属性透传
   */
  className?: string;
  /**
   * `style`属性透传
   */
  style?: CSSProperties;
  /**
   * 根据 `value` 进行比较，判断是否选中
   */
  value?: RadioValue;
  /**
   * 指定当前是否选中
   * @defaultValue false
   */
  checked?: boolean;
  /**
   * 初始是否选中
   * @defaultValue false
   */
  defaultChecked?: boolean;
  /**
   * 禁用
   * @defaultValue false
   */
  disabled?: boolean;
  /**
   * 错误状态
   * @defaultValue false
   */
  error?: boolean;
  /**
   * 自动获取焦点
   * @defaultValue false
   */
  autoFocus?: boolean;
  /**
   * @deprecated 无用
   */
  required?: boolean;
  onMouseEnter?: MouseEventHandler<HTMLLabelElement>;
  onMouseLeave?: MouseEventHandler<HTMLLabelElement>;
  /**
   * 选项变化时的回调函数
   */
  onChange?: (e: RadioChangeEvent) => void;
  /**
   * @internal
   * Radio 样式类型
   */
  optionType?: RadioGroupOptionType;
}

export const Radio = forwardRef<HTMLInputElement, RadioProps>((props, ref) => {
  const {
    prefixCls,
    className,
    role = 'radio',
    checked,
    style,
    onChange,
    onMouseEnter,
    onMouseLeave,
    children,
    value,
    disabled,
    name,
    defaultChecked,
    error,
    optionType = 'default',
    ...restProps
  } = props;
  const context = useRadioGroupContext();
  const inputRef = useRef<HTMLInputElement>(null);
  const cssScope = useCssScope('radio', prefixCls);
  const rootPrefixCls = cssScope();
  const isButton = optionType === 'button';

  const [checkedState, setCheckedState] = useControlledState(
    Boolean(defaultChecked),
    context ? context.value === value : Boolean(checked),
    { isControlledOverride: Boolean(context) || 'checked' in props },
  );

  useImperativeHandle<HTMLInputElement | null, HTMLInputElement | null>(ref, () => inputRef.current);

  function handleChange(e: ChangeEvent<HTMLInputElement>) {
    const eventParams: RadioChangeEvent = {
      target: {
        type: 'radio',
        checked: e.target.checked,
        value,
      },
      stopPropagation() {
        e.stopPropagation();
      },
      preventDefault() {
        e.preventDefault();
      },
      nativeEvent: e.nativeEvent,
      reactEvent: e,
    };

    if (context && isFunction(context.onChange)) {
      context.onChange(eventParams);
    } else {
      setCheckedState(e.target.checked);
      isFunction(onChange) && onChange(eventParams);
    }
  }

  let disableState = disabled;
  let nameState = name;
  let errorState = error;

  if (context) {
    disableState = disabled || context.disabled;
    nameState = context.name;
    errorState = error || context.error;
  }

  const wrapperCls = (() => {
    let cssWrapperSize;
    if (context?.size === 'small') {
      cssWrapperSize = stylex.create({
        padding: '2px 8px',
      });
    } else if (context?.size === 'medium') {
      cssWrapperSize = stylex.create({
        padding: '1px 8px',
      });
    } else if (context?.size === 'large') {
      cssWrapperSize = stylex.create({
        padding: '5px 12px',
      });
    }
    return clsx(
      `${rootPrefixCls}__wrapper`,
      {
        [`${rootPrefixCls}__wrapper--checked`]: checkedState,
        [`${rootPrefixCls}__wrapper--disable`]: disableState,
        [`${rootPrefixCls}--error`]: errorState,
      },
      className,
      stylex(
        cssTypographyBody0,
        // 使用 button style 的时候选中需要加粗
        isButton && checkedState && cssFontMedium,
        isButton && cssWrapperSize,
      ),
    );
  })();

  const sxLabelContent = (() => {
    let cssContentSize: StylexInstance<any> | undefined;
    if (context?.size === 'small') {
      cssContentSize = cssTypographyBody2;
    } else if (context?.size === 'medium') {
      cssContentSize = cssTypographyBody0;
    } else if (context?.size === 'large') {
      cssContentSize = cssTypographyBody0;
    }

    const sxArray: (StylexInstance<any> | undefined)[] = [cssBlockBase];
    if (isButton) {
      sxArray.push(cssContentSize, stylex.create({ maxWidth: RADIO_BUTTON_MAX_WIDTH }));
    } else {
      sxArray.push(
        stylex.create({
          marginLeft: RADIO_LABEL_MARGIN,
          overflowWrap: 'break-word',
        }),
      );
    }

    return sxArray;
  })();

  return (
    <label
      className={wrapperCls}
      style={style}
      onMouseEnter={onMouseEnter}
      onMouseLeave={onMouseLeave}
      data-testid="ud__radio"
    >
      <span className={rootPrefixCls}>
        <input
          ref={inputRef}
          className={clsx(cssScope('__input'))}
          type="radio"
          role={role}
          checked={checkedState}
          onChange={handleChange}
          aria-checked={checkedState}
          disabled={disableState}
          name={nameState}
          {...omit(restProps, ['error'])}
        />
        <span className={cssScope('__wallpaper')} />
        <span className={cssScope('__checked-ink')} />
      </span>
      {children ? (
        <Text
          variant="plain"
          className={clsx(cssScope('__label-content'), disableState && cssScope('__label-content-disabled'))}
          as="span"
          lineClamp={isButton ? 1 : undefined}
          sx={sxLabelContent}
        >
          {children}
        </Text>
      ) : null}
    </label>
  );
});

export default Radio;
