// @universe-design/styles/esm/calendar/index.light.atom.js
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
export const caltoken_addroms_bg_icon_1 = '#e4fae1';
export const caltoken_addroms_bg_icon_2 = '#d0f5ce';
export const caltoken_addroms_bg_icon_3 = '#5cd168';
export const caltoken_btn_selected_line_gray = '#646a73';
export const caltoken_event_border_unanswered_blue = '#336df4';
export const caltoken_event_border_unanswered_carmine = '#df58a5';
export const caltoken_event_border_unanswered_green = '#32a645';
export const caltoken_event_border_unanswered_indigo = '#757df0';
export const caltoken_event_border_unanswered_neutral = '#8f959e';
export const caltoken_event_border_unanswered_orange = '#ed6d0c';
export const caltoken_event_border_unanswered_purple = '#9f6ff1';
export const caltoken_event_border_unanswered_red = '#f54a45';
export const caltoken_event_border_unanswered_tur = '#10a893';
export const caltoken_event_border_unanswered_violet = '#cf5ecf';
export const caltoken_event_border_unanswered_wathet = '#1295ca';
export const caltoken_event_border_unanswered_yellow = '#d99904';
export const caltoken_event_conflict_event_shadow = '8px 8px 24px 8px rgba(31, 35, 41, 0.11), -6px 6px 12px rgba(31, 35, 41, 0.05), -4px 4px 8px -8px rgba(31, 35, 41, 0.08)';
export const caltoken_event_dark_bg_bar_blue = '#336df4';
export const caltoken_event_dark_bg_bar_carmine = '#cc398c';
export const caltoken_event_dark_bg_bar_green = '#258832';
export const caltoken_event_dark_bg_bar_indigo = '#5b65f5';
export const caltoken_event_dark_bg_bar_neutral = 'rgba(100, 106, 115, 0.6)';
export const caltoken_event_dark_bg_bar_orange = '#c25705';
export const caltoken_event_dark_bg_bar_purple = '#8d55ed';
export const caltoken_event_dark_bg_bar_red = '#e22e28';
export const caltoken_event_dark_bg_bar_tur = '#0f8575';
export const caltoken_event_dark_bg_bar_violet = '#bf3dbf';
export const caltoken_event_dark_bg_bar_wathet = '#047fb0';
export const caltoken_event_dark_bg_bar_yellow = '#865b03';
export const caltoken_event_dark_bg_blue = '#5083fb';
export const caltoken_event_dark_bg_carmine = '#df58a5';
export const caltoken_event_dark_bg_green = '#32a645';
export const caltoken_event_dark_bg_indigo = '#757df0';
export const caltoken_event_dark_bg_neutral = '#83888f';
export const caltoken_event_dark_bg_orange = '#ed6d0c';
export const caltoken_event_dark_bg_pending_blue = '#5083fb';
export const caltoken_event_dark_bg_pending_carmine = '#e369ae';
export const caltoken_event_dark_bg_pending_green = '#46a73b';
export const caltoken_event_dark_bg_pending_indigo = '#757df0';
export const caltoken_event_dark_bg_pending_neutral = '#8f959e';
export const caltoken_event_dark_bg_pending_orange = '#ed6d0c';
export const caltoken_event_dark_bg_pending_purple = '#9f6ff1';
export const caltoken_event_dark_bg_pending_red = '#f5605b';
export const caltoken_event_dark_bg_pending_tur = '#10a893';
export const caltoken_event_dark_bg_pending_violet = '#cf5ecf';
export const caltoken_event_dark_bg_pending_wathet = '#1295ca';
export const caltoken_event_dark_bg_pending_yellow = '#ad7a03';
export const caltoken_event_dark_bg_purple = '#9f6ff1';
export const caltoken_event_dark_bg_red = '#f5605b';
export const caltoken_event_dark_bg_tur = '#10a893';
export const caltoken_event_dark_bg_violet = '#cf5ecf';
export const caltoken_event_dark_bg_wathet = '#1295ca';
export const caltoken_event_dark_bg_yellow = '#ad7a03';
export const caltoken_event_dark_calendar_carmine = '#f01d94';
export const caltoken_event_dark_calendar_green = '#2ea121';
export const caltoken_event_dark_calendar_indigo = '#4954e6';
export const caltoken_event_dark_calendar_neutral = '#646a73';
export const caltoken_event_dark_calendar_orange = '#de7802';
export const caltoken_event_dark_calendar_purple = '#7f3bf5';
export const caltoken_event_dark_calendar_red = '#f54a45';
export const caltoken_event_dark_calendar_tur = '#04b49c';
export const caltoken_event_dark_calendar_violet = '#b320b3';
export const caltoken_event_dark_calendar_wathet = '#049fd7';
export const caltoken_event_dark_calendar_yellow = '#dc9b04';
export const caltoken_event_dark_pending_bar_blue = '#336df4';
export const caltoken_event_dark_pending_bar_carmine = '#cc398c';
export const caltoken_event_dark_pending_bar_green = '#1a7526';
export const caltoken_event_dark_pending_bar_indigo = '#5b65f5';
export const caltoken_event_dark_pending_bar_neutral = 'rgba(100, 106, 115, 0.85)';
export const caltoken_event_dark_pending_bar_orange = '#c25705';
export const caltoken_event_dark_pending_bar_purple = '#8d55ed';
export const caltoken_event_dark_pending_bar_red = '#e22e28';
export const caltoken_event_dark_pending_bar_tur = '#0f8575';
export const caltoken_event_dark_pending_bar_violet = '#bf3dbf';
export const caltoken_event_dark_pending_bar_wathet = '#047fb0';
export const caltoken_event_dark_pending_bar_yellow = '#865b03';
export const caltoken_event_detail_bg_left_green = '#5cd168';
export const caltoken_event_detail_bg_left_orange = '#ff9d4c';
export const caltoken_event_detail_bg_left_red = '#ff7570';
export const caltoken_event_detail_bg_left_tur = '#2dbeab';
export const caltoken_event_detail_bg_left_wathet = '#25b0e7';
export const caltoken_event_detail_bg_right_green = '#32a645';
export const caltoken_event_detail_bg_right_orange = '#ed6d0c';
export const caltoken_event_detail_bg_right_red = '#f54a45';
export const caltoken_event_detail_bg_right_tur = '#10a893';
export const caltoken_event_detail_bg_right_wathet = '#1295ca';
export const caltoken_event_light_bg_bar_blue = '#5083fb';
export const caltoken_event_light_bg_bar_carmine = '#df58a5';
export const caltoken_event_light_bg_bar_green = '#32a645';
export const caltoken_event_light_bg_bar_indigo = '#757df0';
export const caltoken_event_light_bg_bar_neutral = '#8f959e';
export const caltoken_event_light_bg_bar_orange = '#ed6d0c';
export const caltoken_event_light_bg_bar_purple = '#9f6ff1';
export const caltoken_event_light_bg_bar_red = '#f54a45';
export const caltoken_event_light_bg_bar_tur = '#10a893';
export const caltoken_event_light_bg_bar_violet = '#cf5ecf';
export const caltoken_event_light_bg_bar_wathet = '#1295ca';
export const caltoken_event_light_bg_bar_yellow = '#d99904';
export const caltoken_event_light_bg_blue = '#e0e9ff';
export const caltoken_event_light_bg_carmine = '#fee2f2';
export const caltoken_event_light_bg_green = '#d0f5ce';
export const caltoken_event_light_bg_indigo = '#e9eafb';
export const caltoken_event_light_bg_neutral = '#dee0e3';
export const caltoken_event_light_bg_oran = '#fee7cd';
export const caltoken_event_light_bg_pending_blue = '#f0f4ff';
export const caltoken_event_light_bg_pending_carmine = '#fef0f8';
export const caltoken_event_light_bg_pending_green = '#e4fae1';
export const caltoken_event_light_bg_pending_indigo = '#f2f3fd';
export const caltoken_event_light_bg_pending_neutral = '#eff0f1';
export const caltoken_event_light_bg_pending_orange = '#fff3e5';
export const caltoken_event_light_bg_pending_purple = '#f5f0fe';
export const caltoken_event_light_bg_pending_red = '#fef0f0';
export const caltoken_event_light_bg_pending_tur = '#e2f8f5';
export const caltoken_event_light_bg_pending_violet = '#fceefc';
export const caltoken_event_light_bg_pending_wathet = '#e7f8fe';
export const caltoken_event_light_bg_pending_yellow = '#fbf4df';
export const caltoken_event_light_bg_purple = '#efe6fe';
export const caltoken_event_light_bg_red = '#fee3e2';
export const caltoken_event_light_bg_tur = '#c4f2ec';
export const caltoken_event_light_bg_violet = '#f9e2f9';
export const caltoken_event_light_bg_wathet = '#caeffc';
export const caltoken_event_light_bg_yellow = '#faedc2';
export const caltoken_event_light_calendar_blue = '#4e83fd';
export const caltoken_event_light_calendar_carmine = '#f14ba9';
export const caltoken_event_light_calendar_green = '#34c724';
export const caltoken_event_light_calendar_indigo = '#616ae5';
export const caltoken_event_light_calendar_neutral = '#9ca2a9';
export const caltoken_event_light_calendar_orange = '#ff8800';
export const caltoken_event_light_calendar_purple = '#935af6';
export const caltoken_event_light_calendar_red = '#f76964';
export const caltoken_event_light_calendar_tur = '#00d6b9';
export const caltoken_event_light_calendar_violet = '#d136d1';
export const caltoken_event_light_calendar_wathet = '#14c0ff';
export const caltoken_event_light_calendar_yellow = '#ffc60a';
export const caltoken_event_light_text_blue = '#0442d2';
export const caltoken_event_light_text_carmine = '#9d1562';
export const caltoken_event_light_text_green = '#0b6017';
export const caltoken_event_light_text_indigo = '#333dcc';
export const caltoken_event_light_text_neutral = '#51565d';
export const caltoken_event_light_text_orange = '#a44904';
export const caltoken_event_light_text_purple = '#4811a6';
export const caltoken_event_light_text_red = '#a11c17';
export const caltoken_event_light_text_tur = '#045d51';
export const caltoken_event_light_text_unanswered_blue = '#3668db';
export const caltoken_event_light_text_unanswered_carmine = '#b14481';
export const caltoken_event_light_text_unanswered_green = '#388c2e';
export const caltoken_event_light_text_unanswered_indigo = '#525ad4';
export const caltoken_event_light_text_unanswered_neutral = '#646a73';
export const caltoken_event_light_text_unanswered_orange = '#b66d36';
export const caltoken_event_light_text_unanswered_purple = '#6d41b8';
export const caltoken_event_light_text_unanswered_red = '#b85451';
export const caltoken_event_light_text_unanswered_tur = '#067062';
export const caltoken_event_light_text_unanswered_violet = '#9f529f';
export const caltoken_event_light_text_unanswered_wathet = '#047fb0';
export const caltoken_event_light_text_unanswered_yellow = '#ad7a03';
export const caltoken_event_light_text_violet = '#872787';
export const caltoken_event_light_text_wathet = '#076a94';
export const caltoken_event_light_text_yellow = '#865b03';
export const caltoken_event_quick_creation_shadow = '0px 10px 36px 10px rgba(31, 35, 41, 0.09), 0px 8px 24px rgba(31, 35, 41, 0.09), 0px 6px 12px -10px rgba(31, 35, 41, 0.11)';
export const caltoken_event_stripe_blue = '#f0f4ff';
export const caltoken_event_stripe_carmine = '#fef0f8';
export const caltoken_event_stripe_green = '#e4fae1';
export const caltoken_event_stripe_indigo = '#f2f3fd';
export const caltoken_event_stripe_neutral = '#f2f3f5';
export const caltoken_event_stripe_orange = '#fff3e5';
export const caltoken_event_stripe_purple = '#f5f0fe';
export const caltoken_event_stripe_red = '#fef0f0';
export const caltoken_event_stripe_tur = '#e2f8f5';
export const caltoken_event_stripe_violet = '#fceefc';
export const caltoken_event_stripe_wathet = '#e7f8fe';
export const caltoken_event_stripe_yellow = '#fbf4df';
export const caltoken_tag_colour_day = '#e29b00';
