// @byted-service/bytedtrace/lib/event/event.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TraceEvent = void 0;
const tslib_1 = require("tslib");
const gConsts = tslib_1.__importStar(require("../consts"));
const bytedtracePb = tslib_1.__importStar(require("../gen/bytedtrace_pb"));
class TraceEvent {
    constructor(options) {
        // for keeping original file location in log
        /**
         * 记录调用层级，给 log 定位
         */
        this.callDepth = 0;
        this.eventType = options.eventType;
        this.tags = options.tags;
        this.name = options.name;
        this.logLevel = options.logLevel;
        this.content = options.content || '';
    }
    /**
     * 返回 event 发生时间
     * @returns
     */
    getStartTime() {
        return this.startTime;
    }
    /**
     * 返回 event 描述内容
     * @returns
     */
    getContent() {
        return this.content;
    }
    /**
     * 返回 event 类型
     * @returns
     */
    getEventType() {
        return this.eventType;
    }
    /**
     * 返回 tags 内容，这里返回新的对象
     * @returns
     */
    getTags() {
        return Object.assign({}, this.tags);
    }
    /**
     * 返回 log level
     * @returns
     */
    getLogLevel() {
        return this.logLevel;
    }
    /**
     * 返回 event 名字
     * @returns
     */
    getName() {
        return this.name;
    }
    /**
     * 返回是否打印 log 的判断标志值
     * @returns
     */
    getEmitLog() {
        return this.emitLog;
    }
    /**
     * 设置一对 key-value 到 tags 中
     * @param key
     * @param value
     * @returns
     */
    setTag(key, value) {
        this.tags[key] = value;
        return this;
    }
    /**
     * 设置多个 key-value 到 tags 中
     * @param tags
     * @returns
     */
    setTags(tags) {
        const keys = Object.keys(tags);
        for (let i = 0; i < keys.length; i++) {
            const key = keys[i];
            if (tags[key] !== undefined) {
                this.tags[key] = tags[key];
            }
        }
        return this;
    }
    /**
     * 设置是否打印 log
     * @param e
     * @returns
     */
    setEmitLog(e) {
        this.emitLog = e;
        if (this.logLevel === gConsts.LogLevels.LEVEL_NO_DEFINE) {
            this.logLevel = gConsts.LogLevels.LEVEL_INFO;
        }
        return this;
    }
    /**
     * 增加 callDepth
     * @param depth
     * @returns
     */
    incCallDepthForLog(depth) {
        this.callDepth += depth;
        return this;
    }
    /**
     * 返回 CallDepth 值
     * @returns
     */
    getCallDepth() {
        return this.callDepth;
    }
    /**
     * 设置是否进行 metrics 打点
     * @param e
     * @returns
     */
    setEmitMetrics(e) {
        this.emitMetrics = e;
        return this;
    }
    /**
     * 获取是否进行 metrics 打点的状态值
     * @returns
     */
    getEmitMetrics() {
        return this.emitMetrics;
    }
    /**
     * 设置描述内容
     * @param content
     * @returns
     */
    setContent(content) {
        this.content = content;
        return this;
    }
    /**
     * 设置发生时间戳
     * @param time
     * @returns
     */
    setTimestamp(time) {
        this.startTime = time;
        return this;
    }
    tranToPb(breaker) {
        const eventPb = new bytedtracePb.EventProto();
        eventPb.setName(this.name);
        eventPb.setType(this.eventType);
        eventPb.setData(breaker.truncatedStringToBytesContent(this.content));
        eventPb.setStartTime(this.startTime);
        const tags = [];
        const keys = Object.keys(this.tags);
        for (let i = 0; i < keys.length; i++) {
            const key = keys[i];
            const val = this.tags[key];
            tags.push([key, val]);
        }
        eventPb.setTagsList(breaker.buildTags(tags) || []);
        if (this.logLevel !== gConsts.LogLevels.LEVEL_NO_DEFINE) {
            eventPb.addTags(breaker.buildTag(gConsts.LOG_LEVEL_KEY, this.logLevel));
        }
        return eventPb;
    }
}
exports.TraceEvent = TraceEvent;
