// @byted-service/netpool/lib/internal/pending_promise.js
"use strict";
/* tslint:disable max-classes-per-file */
/* eslint-disable max-classes-per-file */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PendingPromise = void 0;
const errors_1 = require("./errors");
const onTimeout = (p) => {
    p.reject(p.possibleTimeoutCause
        ? new errors_1.FakeError(p.possibleTimeoutCause)
        : new errors_1.TimeoutError(`${p.type} wait timeout ${p.timeout}ms`));
};
class PendingPromise {
    constructor(type, timeout) {
        this.isFinaled = false;
        this.possibleTimeoutCause = null;
        this.timeoutId = null;
        this.promise = new Promise((resolve, reject) => {
            this._resolve = resolve;
            this._reject = reject;
        });
        this.type = type;
        this.timeout = timeout;
        if (typeof this.timeout === 'number') {
            this.timeoutId = setTimeout(onTimeout, this.timeout, this);
        }
    }
    resolve(value) {
        if (this.isFinaled) {
            return this;
        }
        if (this.timeoutId) {
            clearTimeout(this.timeoutId);
            this.timeoutId = null;
        }
        this.isFinaled = true;
        this.value = value;
        this._resolve(this);
        return this;
    }
    reject(reason) {
        if (this.isFinaled) {
            return this;
        }
        if (this.timeoutId) {
            clearTimeout(this.timeoutId);
            this.timeoutId = null;
        }
        this.isFinaled = true;
        this.reason = reason;
        this._reject(this);
        return this;
    }
}
exports.PendingPromise = PendingPromise;
