// @byted-service/netpool/lib/internal/pool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Pool = void 0;
const util_1 = require("util");
const errors_1 = require("./errors");
const pending_promise_1 = require("./pending_promise");
const helper_1 = require("../helper");
const consts_1 = require("../consts");
const eventIdSym = Symbol.for('#eventId');
const resourceSym = Symbol.for('#resource');
const pSym = Symbol.for('#p');
const sleep = (0, util_1.promisify)(setTimeout);
const idCreator = (0, helper_1.genIdCreator)('pool');
/**
 * https://bytedance.feishu.cn/docs/doccnFk064LQeLPglQGkj5LMi7b#
 */
class Pool {
    constructor(options) {
        this.version = consts_1.version;
        this.creating = new Set();
        /** FIFI: -> */
        this.idling = [];
        this.using = new Set();
        this.destroying = new Set();
        /** FIFI: -> */
        this.acquireQueue = [];
        this.createQueue = new Set();
        this.useQueue = new Set();
        this.destroyQueue = new Set();
        this.resourceToPromise = new Map();
        this.destroyed = false;
        this.eventId = 1;
        this.eventMap = new Map();
        this.reapIntervalId = null;
        this.onAcquireSuccess = (p) => {
            const eventId = p[eventIdSym];
            const value = p.value;
            this.emit('acquireSuccess', eventId, value);
            this.doReset(value, 'acquire');
            this.doUse(value);
            return value;
        };
        this.onAcquireFail = (p) => {
            const eventId = p[eventIdSym];
            const reason = p.reason;
            this.emit('acquireFail', eventId, reason);
            return Promise.reject(reason);
        };
        this.onCreateSuccess = (p) => {
            const eventId = p[eventIdSym];
            const resource = p[resourceSym];
            const value = p.value;
            /**
             * Set the value back to resource to supplement relevant information when debugging.
             *
             * 将值设置回 resource，用于 debug 时补充相关信息
             */
            if (value) {
                const keys = Object.keys(resource);
                if (keys.length > 0) {
                    for (let i = 0; i < keys.length; i++) {
                        value[keys[i]] = resource[keys[i]];
                    }
                }
            }
            this.emit('createSuccess', eventId, value);
            this.creating.delete(resource);
            this.idling.unshift(value);
            this.createQueue.delete(p);
            this.startReaping();
            this.triggerAcquires();
            return value;
        };
        this.onCreateFail = (p) => {
            const eventId = p[eventIdSym];
            const resource = p[resourceSym];
            const reason = p.reason;
            this.emit('createFail', eventId, reason);
            this.creating.delete(resource);
            this.createQueue.delete(p);
            /**
             * Save the create error to all pending acquires so that we can use it
             * as the error to reject the acquire if it times out.
             *
             * 保存错误，作为可能的超时原因
             */
            const ps = this.acquireQueue;
            for (let i = 0; i < ps.length; i++) {
                ps[i].possibleTimeoutCause = reason;
            }
            if (this.throwOnFirstCreateError && p[pSym]) {
                p[pSym].reject(reason);
            }
            this.createRetryInterval
                ? sleep(this.createRetryInterval).then(() => this.triggerAcquires())
                : this.triggerAcquires();
        };
        this.onUseSuccess = (p) => {
            const resource = p[resourceSym];
            this.using.delete(resource);
            this.idling.unshift(resource);
            this.useQueue.delete(p);
            this.resourceToPromise.delete(resource);
            this.triggerAcquires();
        };
        this.onUseFail = (p) => {
            const resource = p[resourceSym];
            this.emit('useTimeout', resource, p.timeout);
            this.using.delete(resource);
            this.useQueue.delete(p);
            this.resourceToPromise.delete(resource);
            this.triggerAcquires();
        };
        this.onDestroySuccess = (p) => {
            const eventId = p[eventIdSym];
            const resource = p[resourceSym];
            this.emit('destroySuccess', eventId, resource);
            this.destroying.delete(resource);
            this.destroyQueue.delete(p);
        };
        this.onDestroyFail = (p) => {
            const eventId = p[eventIdSym];
            const resource = p[resourceSym];
            const reason = p.reason;
            this.emit('destroyFail', eventId, resource, reason);
            this.destroying.delete(resource);
            this.destroyQueue.delete(p);
        };
        this.id = options.id || idCreator();
        this.max =
            typeof options.max === 'number'
                ? options.max
                : consts_1.defaultPoolOptions.max;
        this.min =
            typeof options.min === 'number'
                ? options.min
                : consts_1.defaultPoolOptions.min;
        this.createRetryInterval =
            typeof options.createRetryInterval === 'number'
                ? options.createRetryInterval
                : consts_1.defaultPoolOptions.createRetryInterval;
        this.reapInterval =
            options.reapInterval || consts_1.defaultPoolOptions.reapInterval;
        this.throwOnFirstCreateError =
            typeof options.throwOnFirstCreateError === 'boolean'
                ? options.throwOnFirstCreateError
                : consts_1.defaultPoolOptions.throwOnFirstCreateError;
        this.concurrencyTolerance =
            typeof options.concurrencyTolerance === 'number'
                ? options.concurrencyTolerance
                : consts_1.defaultPoolOptions.concurrencyTolerance;
        this.maxConcurrencyCreating =
            options.maxConcurrencyCreating ||
                this.max + this.max * this.concurrencyTolerance;
        this.operation = options.operation;
        this.logger = options.logger || consts_1.defaultLogger;
        this.acquireTimeout =
            typeof options.acquireTimeout === 'number' ||
                options.acquireTimeout === null
                ? options.acquireTimeout
                : consts_1.defaultPoolOptions.acquireTimeout;
        this.createTimeout =
            typeof options.createTimeout === 'number' ||
                options.createTimeout === null
                ? options.createTimeout
                : consts_1.defaultPoolOptions.createTimeout;
        this.useTimeout =
            typeof options.useTimeout === 'number' ||
                options.useTimeout === null
                ? options.useTimeout
                : consts_1.defaultPoolOptions.useTimeout;
        this.destroyTimeout =
            typeof options.destroyTimeout === 'number' ||
                options.destroyTimeout === null
                ? options.destroyTimeout
                : consts_1.defaultPoolOptions.destroyTimeout;
        if (this.useTimeout === null) {
            this.release = function release(resource) {
                this.emit('release', resource);
                if (this.using.has(resource)) {
                    this.doReset(resource, 'release');
                    this.using.delete(resource);
                    this.idling.unshift(resource);
                    this.triggerAcquires();
                    return true;
                }
                this.doDestroy(resource);
                return false;
            };
            this.doUse = function doUse(resource) {
                this.using.add(resource);
            };
        }
    }
    async acquire(options) {
        const timeout = typeof options === 'number' || options === null
            ? options
            : this.acquireTimeout;
        const eventId = this.eventId++;
        this.emit('acquireRequest', eventId, timeout);
        /**
         * If there are free resources, execute synchronous acquisition logic
         * 如果存在空闲资源，就执行同步获取逻辑
         */
        if (this.idling.length > 0) {
            let resource;
            while ((resource = this.idling.pop())) {
                if (this.doValidate(resource)) {
                    this.emit('acquireSuccess', eventId, resource);
                    this.doReset(resource, 'acquire');
                    this.doUse(resource);
                    return resource;
                }
                this.doDestroy(resource);
            }
        }
        const p = new pending_promise_1.PendingPromise('acquiring', timeout);
        p[eventIdSym] = eventId;
        this.acquireQueue.unshift(p);
        this.triggerAcquires();
        return p.promise.then(this.onAcquireSuccess, this.onAcquireFail);
    }
    release(resource) {
        this.emit('release', resource);
        const p = this.resourceToPromise.get(resource);
        if (p) {
            this.doReset(resource, 'release');
            p.resolve();
            return true;
        }
        this.doDestroy(resource);
        return false;
    }
    async destroy() {
        if (this.destroyed) {
            return;
        }
        const eventId = this.eventId++;
        this.emit('poolDestroyRequest', eventId);
        this.destroyed = true;
        this.acquire = async function acquire() {
            throw new Error('Pool already destroyed');
        };
        this.stopReaping();
        if (this.createQueue.size > 0) {
            await sleep(0);
            await Promise.all(Array.from(this.createQueue.keys()).map((p) => new Promise((resolve) => p.promise.then(resolve).catch(resolve))));
        }
        if (this.useQueue.size > 0) {
            await sleep(0);
            await Promise.all(Array.from(this.useQueue.keys()).map((p) => new Promise((resolve) => p.promise.then(resolve).catch(resolve))));
        }
        if (this.acquireQueue.length > 0) {
            await sleep(0);
            const ps = this.acquireQueue;
            for (let i = 0; i < ps.length; i++) {
                ps[i].reject(new errors_1.AbortError('Pool is destroying'));
            }
        }
        for (const resource of [...this.idling, ...this.using]) {
            this.doDestroy(resource);
        }
        if (this.destroyQueue.size > 0) {
            await Promise.all(Array.from(this.destroyQueue.keys()).map((p) => new Promise((resolve) => p.promise.then(resolve).catch(resolve))));
        }
        this.creating.clear();
        this.idling = [];
        this.using.clear();
        this.destroying.clear();
        this.acquireQueue = [];
        this.createQueue.clear();
        this.useQueue.clear();
        this.destroyQueue.clear();
        this.resourceToPromise.clear();
        this.emit('poolDestroySuccess', eventId);
        this.eventId = 1;
        this.eventMap.clear();
    }
    stats() {
        return {
            id: this.id,
            version: this.version,
            timestamp: Date.now(),
            destroyed: this.destroyed,
            total: this.creating.size + this.idling.length + this.using.size,
            creating: this.creating.size,
            idling: this.idling.length,
            using: this.using.size,
            destroying: this.destroying.size,
            acquireQueue: this.acquireQueue.length,
            createQueue: this.createQueue.size,
            useQueue: this.useQueue.size,
            destroyQueue: this.destroyQueue.size,
        };
    }
    getResources() {
        return {
            creating: Array.from(this.creating.keys()),
            idling: Array.from(this.idling),
            using: Array.from(this.using.keys()),
            destroying: Array.from(this.destroying.keys()),
        };
    }
    on(event, listener) {
        const listeners = this.eventMap.get(event) || [];
        listeners.push(listener);
        this.eventMap.set(event, listeners);
    }
    emit(event, ...args) {
        let listeners = this.eventMap.get(event);
        if (!listeners || listeners.length === 0) {
            return;
        }
        listeners = listeners.slice(0);
        for (let i = 0; i < listeners.length; i++) {
            try {
                listeners[i](...args);
            }
            catch (err) {
                this.logger.warn('[netpool] [%s] Failed to emit listener for event "%s". %o', this.id, event, err);
            }
        }
    }
    triggerAcquires() {
        const acquireQueue = this.acquireQueue;
        if (acquireQueue.length === 0) {
            return;
        }
        const creating = this.creating;
        const idling = this.idling;
        const using = this.using;
        /**
         * In certain momentary high concurrency scenarios, the momentary connection pool may run out and there are still a large number of applications waiting
         * 在某些瞬间高并发的场景下，可能会出现瞬间连接池用完并且还有大量申请等待的情况
         *
         * The concept of concurrency tolerance is introduced here: when there are a large number of applications, the number of connections in the connection pool is allowed to be greater than options.max within a certain period of time, and then it will try to recover in reapInterval
         * 这里引入并发容忍度这一概念：当出现大量申请时，允许在一定时间内连接池内连接数量大于 options.max，之后将会在 reapInterval 中尝试进行恢复
         */
        const max = this.max + this.max * this.concurrencyTolerance;
        /**
         * Resource exhaustion
         * 资源耗尽
         */
        if (creating.size + using.size >= max) {
            return;
        }
        if (idling.length === 0) {
            /**
             * Concurrent creation limit exceeded
             * 超过并发创建限制
             */
            if (creating.size >= this.maxConcurrencyCreating) {
                return;
            }
        }
        while (true) {
            const p = acquireQueue.pop();
            if (!p) {
                break;
            }
            if (p.isFinaled) {
                continue;
            }
            const resource = idling.pop();
            if (!resource) {
                /**
                 * Resource exhaustion
                 * 资源耗尽
                 */
                if (creating.size + using.size >= max) {
                    acquireQueue.push(p);
                    break;
                }
                /**
                 * Concurrent creation limit exceeded
                 * 超过并发创建限制
                 */
                if (creating.size >= this.maxConcurrencyCreating) {
                    acquireQueue.push(p);
                    break;
                }
                /**
                 * Exclude creating
                 * 排除创建中
                 */
                if (creating.size > acquireQueue.length) {
                    acquireQueue.push(p);
                    break;
                }
                this.doCreate()[pSym] = this.throwOnFirstCreateError ? p : null;
                acquireQueue.push(p);
                break;
            }
            if (this.doValidate(resource)) {
                /**
                 * Because doUse here is asynchronous, it is added to using in advance to avoid too many creations
                 * 因为这里的 doUse 是异步，所以提前添加到 using 中，避免过多创建
                 */
                using.add(resource);
                p.resolve(resource);
                continue;
            }
            this.doDestroy(resource);
            acquireQueue.push(p);
        }
    }
    startReaping() {
        if (this.destroyed) {
            return;
        }
        if (!this.reapIntervalId) {
            this.emit('startReaping', this.reapInterval);
            this.reapIntervalId = setInterval(() => {
                const creating = this.creating;
                const idling = this.idling;
                const using = this.using;
                const temp = [];
                let resource;
                while ((resource = idling.pop())) {
                    if (this.doValidate(resource)) {
                        temp.push(resource);
                        continue;
                    }
                    this.doDestroy(resource);
                }
                idling.push(...temp);
                while (idling.length > this.max) {
                    const resource = idling.shift();
                    this.emit('destroyExcessResource', resource);
                    this.doDestroy(resource);
                }
                while (creating.size + idling.length + using.size < this.min) {
                    /**
                     * Concurrent creation limit exceeded
                     * 超过并发创建限制
                     */
                    if (creating.size >= this.maxConcurrencyCreating) {
                        break;
                    }
                    this.doCreate();
                }
                if (creating.size + idling.length + using.size === 0) {
                    this.stopReaping();
                }
            }, this.reapInterval).unref();
        }
    }
    stopReaping() {
        if (this.reapIntervalId) {
            this.emit('stopReaping');
            clearInterval(this.reapIntervalId);
        }
        this.reapIntervalId = null;
    }
    doCreate() {
        const timeout = this.createTimeout;
        const eventId = this.eventId++;
        this.emit('createRequest', eventId, timeout);
        const resource = {};
        const p = new pending_promise_1.PendingPromise('creating', timeout);
        p[eventIdSym] = eventId;
        p[resourceSym] = resource;
        this.creating.add(resource);
        this.createQueue.add(p);
        p.promise.then(this.onCreateSuccess, this.onCreateFail);
        this.operation
            .create(timeout)
            .then(p.resolve.bind(p), p.reject.bind(p));
        return p;
    }
    doValidate(resource) {
        return this.operation.validate(resource);
    }
    doUse(resource) {
        const timeout = this.useTimeout;
        const p = new pending_promise_1.PendingPromise('using', timeout);
        p[resourceSym] = resource;
        this.using.add(resource);
        this.useQueue.add(p);
        this.resourceToPromise.set(resource, p);
        p.promise.then(this.onUseSuccess, this.onUseFail);
    }
    doDestroy(resource) {
        const timeout = this.destroyTimeout;
        const eventId = this.eventId++;
        this.emit('destroyRequest', eventId, resource, timeout);
        const p = new pending_promise_1.PendingPromise('destroying', timeout);
        p[eventIdSym] = eventId;
        p[resourceSym] = resource;
        this.destroying.add(resource);
        this.destroyQueue.add(p);
        p.promise.then(this.onDestroySuccess, this.onDestroyFail);
        this.operation
            .destroy(resource, timeout)
            .then(p.resolve.bind(p), p.reject.bind(p));
        return p;
    }
    doReset(resource, stage) {
        this.operation.reset(resource, stage);
    }
}
exports.Pool = Pool;
