// @byted-service/logger/lib/streams/file_rotate/file_rotate_stream.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileRotateStream = void 0;
const tslib_1 = require("tslib");
const fs_1 = tslib_1.__importDefault(require("fs"));
const util_1 = require("util");
const sonic_boom_1 = tslib_1.__importDefault(require("sonic-boom"));
const dayjs_1 = tslib_1.__importDefault(require("dayjs"));
const cron_expression_1 = require("./cron_expression");
const consts_1 = require("../../consts");
const defaultRotatePattern = '0 0 * * *'; // 每小时切割一次
class FileRotateStream {
    constructor(options) {
        this.rotateIntervalId = null;
        // 为了单测
        this.rename = (0, util_1.promisify)(fs_1.default.rename);
        this.file = options.file;
        this.sonicBoom = new sonic_boom_1.default({
            fd: this.file,
            minLength: 0,
            sync: false,
        });
        this.sonicBoom.setMaxListeners(1000);
        /* istanbul ignore next */
        const filterBrokenPipe = (err) => {
            if (err.code === 'EPIPE') {
                // If we get EPIPE, we should stop logging here
                // however we have no control to the consumer of
                // SonicBoom, so we just overwrite the write method
                this.sonicBoom.write = consts_1.noop;
                this.sonicBoom.end = consts_1.noop;
                this.sonicBoom.flushSync = consts_1.noop;
                this.sonicBoom.destroy = consts_1.noop;
                console.warn('[logger] Got EPIPE while write log to file "%s". It is impossible to write again, the process should be shut down', this.file);
                return;
            }
            this.sonicBoom.removeListener('error', filterBrokenPipe);
        };
        this.sonicBoom.on('error', filterBrokenPipe);
        this.updateRotate(options);
    }
    updateRotate(options) {
        this.rotatePattern =
            options.rotatePattern || this.rotatePattern || defaultRotatePattern;
        this.fileFormatter = options.fileFormatter || this.fileFormatter;
        this.expression = new cron_expression_1.CornExpression(this.rotatePattern);
        if (this.rotateIntervalId) {
            clearTimeout(this.rotateIntervalId);
        }
        const rotateFile = () => {
            this.rotateIntervalId = setTimeout(async () => {
                await this.rotate();
                rotateFile();
            }, this.expression.next().toDate().getTime() - Date.now()).unref();
        };
        rotateFile();
    }
    write(msg) {
        this.sonicBoom.write(msg);
    }
    flush() {
        this.sonicBoom.flush();
    }
    flushSync() {
        this.sonicBoom.flushSync();
    }
    reopen() {
        this.sonicBoom.reopen(this.file);
    }
    destroy() {
        if (this.rotateIntervalId) {
            clearTimeout(this.rotateIntervalId);
        }
        this.rotateIntervalId = null;
        if (!this.sonicBoom.destroyed) {
            this.sonicBoom.flush();
            this.sonicBoom.end();
        }
    }
    async rotate() {
        try {
            const prev = this.expression.prev();
            const rotateFile = this.fileFormatter
                ? this.fileFormatter(this.file, prev.toDate())
                : `${this.file}.${prev.tz().format('YYYY-MM-DD_HH:mm:ss')}`;
            if (!fs_1.default.existsSync(rotateFile)) {
                await this.rename(this.file, rotateFile);
            }
        }
        catch (err) {
            if (typeof err.message !== 'string' ||
                err.message.indexOf('no such file or directory') < 0) {
                console.log(`[logger] Failed to rotate file "%s" at "%s", don\'t worry, it will try at next time.`, this.file, (0, dayjs_1.default)().tz().format('YYYY-MM-DD HH:mm:ss'), err);
            }
        }
        finally {
            this.reopen();
        }
    }
}
exports.FileRotateStream = FileRotateStream;
