// @universe-design/react/src/dropdown/Button.tsx
import type { ReactElement } from 'react';
import React from 'react';
import type { ButtonProps } from '../button';
import { Button } from '../button';
import type { DropdownProps } from './Dropdown';
import { Dropdown } from './Dropdown';
import MoreOutlined from '@universe-design/icons-react/MoreOutlined';
import { useCssScope } from '../config-provider/hooks';
import clsx from 'clsx';

import './style';

/**
 * @public
 * @docgen
 * <Dropdown.Button> 的属性和方法。
 */
export interface DropdownButtonProps
  extends Omit<DropdownProps, 'children'>,
    Pick<ButtonProps, 'size' | 'type' | 'htmlType' | 'icon' | 'href' | 'title'>,
    React.PropsWithChildren<unknown> {
  // ButtonGroupProps
  /**
   * 按钮组的 `style` 属性透传
   */
  style?: React.CSSProperties;
  /**
   * 按钮组的 `className` 属性透传
   */
  className?: string;
  /**
   * 左按钮的点击事件回调
   */
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
  /**
   * 按钮组的自定义渲染函数
   */
  buttonsRender?: (buttons: [React.ReactNode, React.ReactElement]) => [React.ReactNode, React.ReactElement];
}

export const DropdownButton: React.FC<DropdownButtonProps> = props => {
  const {
    type,
    disabled,
    htmlType,
    children,
    overlay,
    trigger,
    align,
    visible,
    defaultVisible,
    placement = 'bottomRight',
    href,
    icon = <MoreOutlined />,
    title,
    className,
    style,
    getPopupContainer,
    onVisibleChange,
    buttonsRender,
    onClick,
    size,
    ...restProps
  } = props;

  const cssScope = useCssScope('dropdown-button');

  const dropdownProps = {
    align,
    overlay,
    disabled,
    visible,
    defaultVisible,
    placement,
    trigger,
    onVisibleChange,
    getPopupContainer,
    ...restProps,
  } as DropdownProps;

  const leftButton = (
    <Button size={size} type={type} disabled={disabled} onClick={onClick} htmlType={htmlType} href={href} title={title}>
      {children}
    </Button>
  );

  const rightButton: ReactElement = <Button size={size} disabled={disabled} type={type} shape="square" icon={icon} />;

  const [leftButtonToRender, rightButtonToRender] = buttonsRender
    ? buttonsRender([leftButton, rightButton])
    : [leftButton, rightButton];

  return (
    <div style={style} className={clsx(cssScope(), className)}>
      {leftButtonToRender}
      <Dropdown {...dropdownProps}>{rightButtonToRender}</Dropdown>
    </div>
  );
};
