// @universe-design/react/src/dropdown/Dropdown.tsx
/* eslint-disable max-lines-per-function */
import type { ReactElement } from 'react';
import React, { forwardRef, useCallback, useMemo, useRef } from 'react';
import { useCssScope } from '../config-provider/hooks';
import { usePopperSidecar } from '../config-provider/hooks/usePopperSidecar';
import type { PositionType } from '../util';
import { isFunction } from '../util';
import type { RcMenuInfo } from '../menu';
import type { UDAlignType, UdTriggerRef } from '../shared/components/trigger';
import { DropdownMenuContext } from './DropdownMenuContext';
import composeRefs from '../util/composeRefs';
import type { PopperProps, PopperRef } from '../shared/components/popper';
import { SlideUpDownTransition } from '../shared/components/transition/SlideUpDownTransition';

import './style';

/**
 * @public
 */
export type DropdownPlacementType = Exclude<PositionType, 'leftCenter' | 'rightCenter'>;

/**
 * @public
 * @docgen
 */
export interface DropdownProps
  extends Pick<
    PopperProps,
    | 'getPopupContainer'
    | 'mouseEnterDelay'
    | 'mouseLeaveDelay'
    | 'showAction'
    | 'hideAction'
    | 'disabled'
    | 'modifiers'
    | 'destroyOnHide'
    | 'visible'
    | 'defaultVisible'
    | 'onVisibleChange'
    | 'openClassName'
    | 'udTriggerRef'
    | 'trigger'
    // new popper props
    | 'offset'
    | 'popupPortalContainer'
    | 'flip'
    | 'shift'
    | 'enableAutoHide'
    | 'eventListeners'
    | 'enableMaxHeight'
    | 'middlewares'
    | 'inline'
    | 'boundary'
    | 'initialFocus'
    | 'enableFocusLock'
    | 'enableRestoreFocus'
    | 'boundaryPadding'
  > {
  /**
   * 点击弹层事件回调
   */
  onOverlayClick?: (ev: RcMenuInfo) => void;
  prefixCls?: string;
  /**
   * 弹层的 `className` 属性透传
   */
  overlayClassName?: string;
  /**
   *  弹层的 `style` 属性透传
   */
  overlayStyle?: React.CSSProperties;
  /**
   * 完全自定义对齐规则，会与内建对齐规则合并
   */
  align?: UDAlignType;
  /**
   * 内置方位
   */
  placement?: DropdownPlacementType;
  /**
   * 弹层。`overlay`菜单使用：[导航菜单 Navigation Menu](/develop/react/menu)
   */
  overlay: (() => React.ReactElement) | React.ReactElement;
  children: ReactElement;
}

// eslint-disable-next-line max-lines-per-function
export const Dropdown = forwardRef<PopperRef, DropdownProps>((props, ref) => {
  const {
    prefixCls: prefixClsInProps,
    align: alignInProps,
    placement = 'bottomLeft',
    showAction,
    hideAction,
    overlayClassName,
    overlayStyle,
    trigger = 'hover',
    disabled = false,
    onOverlayClick,
    overlay,
    children,
    udTriggerRef: udTriggerRefProp,
    ...otherProps
  } = props;
  const cssScope = useCssScope('dropdown');
  const Popper = usePopperSidecar();
  const prefixCls = prefixClsInProps || cssScope();
  const udTriggerRef = useRef<UdTriggerRef>(null);

  const mergedUdTriggerRef = React.useMemo(
    () => composeRefs(udTriggerRef, udTriggerRefProp),
    [udTriggerRef, udTriggerRefProp],
  );

  const overlayNode = useMemo((): React.ReactElement => (isFunction(overlay) ? overlay() : overlay), [overlay]);

  const onClick = useCallback(
    (e: RcMenuInfo) => {
      isFunction(onOverlayClick) && onOverlayClick(e);
      if (e && 'domEvent' in e && e.domEvent.isDefaultPrevented()) {
        return;
      }
      udTriggerRef.current?.setVisible(false);
    },
    [onOverlayClick],
  );

  const renderOverlay = (
    <DropdownMenuContext.Provider
      value={useMemo(
        () => ({
          className: `${prefixCls}__overlay`,
          onClick,
        }),
        [onClick, prefixCls],
      )}
    >
      {overlayNode}
    </DropdownMenuContext.Provider>
  );

  const alignPoint = trigger.indexOf('contextMenu') !== -1;

  return (
    <Popper
      {...otherProps}
      disabled={disabled}
      alignPoint={alignPoint}
      prefixCls={prefixCls}
      ref={ref}
      udTriggerRef={mergedUdTriggerRef}
      popupClassName={overlayClassName}
      popupStyle={overlayStyle}
      action={trigger}
      showAction={showAction}
      hideAction={hideAction}
      popupPlacement={placement}
      popupAlign={alignInProps}
      popupMotion="slide-up"
      popup={renderOverlay}
      {...{
        showArrow: false,
        arrowHasStroke: false,
        Transition: SlideUpDownTransition,
      }}
      defaultMainAxisOffsetWithoutArrow={4}
      __screenshotAlwaysOpenEnabled
    >
      {React.useCallback(
        ({ visible }) =>
          'selected' in children.props ? children : React.cloneElement(children, { selected: visible }),
        [children],
      )}
    </Popper>
  );
});
if (__DEV__) {
  Dropdown.displayName = 'Dropdown';
}
