// @universe-design/react/lib/switch/Switch.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Switch = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _clsx = _interopRequireDefault(require("clsx"));
var _react = _interopRequireDefault(require("react"));
var _omit = _interopRequireDefault(require("lodash/omit"));
var _reactUtilities = require("@universe-design/react-utilities");
var _util = require("../util");
var _hooks = require("../config-provider/hooks");
var _svgs = require("../_assets/svgs");
var _forwardRef = require("../util/forwardRef");
require("./style");
const Switch = exports.Switch = /* #__PURE__*/Object.assign((0, _forwardRef.forwardHTMLRef)(function (props, ref) {
  const ctxSize = (0, _hooks.useComponentSize)();
  const {
    checked: checkedProp,
    defaultChecked = false,
    loading = false,
    disabled: disabledProp = false,
    size = ctxSize === 'large' ? 'medium' : ctxSize,
    onChange,
    onClick,
    className,
    style,
    id,
    title,
    autoFocus,
    unCheckedChildren,
    checkedChildren,
    prefixCls,
    ...restProp
  } = props;
  process.env.NODE_ENV !== "production" ? (0, _util.warning)(!unCheckedChildren && !checkedChildren, (0, _util.devWarning)(Switch.displayName || Switch.name, 'unCheckedChildren and checkedChildren are not support.')) : void 0;
  process.env.NODE_ENV !== "production" ? (0, _util.warning)('checked' in props || !('value' in props), (0, _util.devWarning)('Switch', '`value` is not a valid prop, do you mean `checked`?')) : void 0;
  const cssScope = (0, _hooks.useCssScope)('switch', prefixCls);
  const [checked, setChecked] = (0, _reactUtilities.useControlledState)(checkedProp !== null && checkedProp !== void 0 ? checkedProp : defaultChecked, checkedProp);
  const disabled = disabledProp || loading;
  const usedSize = (0, _util.getSizeAbbr)(size);
  return /*#__PURE__*/_react.default.createElement("button", (0, _extends2.default)({}, (0, _omit.default)(restProp, 'error'), {
    onClick: _react.default.useCallback(e => {
      if (!disabled) {
        setChecked(_checked => !_checked, _checked => {
          onChange === null || onChange === void 0 ? void 0 : onChange(_checked, e);
          onClick === null || onClick === void 0 ? void 0 : onClick(_checked, e);
        });
      }
    }, [disabled, onChange, onClick, setChecked]),
    className: (0, _clsx.default)(cssScope(), checked && cssScope('-checked'), disabled && cssScope('-disabled'), className, {
      [cssScope(`-${usedSize}`)]: usedSize
    }),
    style: style,
    id: id,
    type: "button",
    role: "switch",
    autoFocus: autoFocus,
    title: title,
    ref: ref,
    "aria-checked": checked
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: cssScope('__handler')
  }, loading ? /*#__PURE__*/_react.default.createElement(_svgs.SpinLoadingSvg, {
    className: cssScope('__loading')
  }) : null));
}), {
  __UD_SWITCH: true
});

/**
 * @docgen
 */

if (process.env.NODE_ENV !== "production") {
  Switch.displayName = 'Switch';
}
