// @universe-design/react/es/message/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import React, { useRef } from 'react';
import ReactDOM from 'react-dom';
import { CrossRootConfigProvider, ConfigProvider } from '../config-provider';
import { Toast } from './Toast';
import './style';
let globalToast;
let globalToastInitPromise;
function getDefaultContainer() {
  const div = document.createElement('div');
  document.body.appendChild(div);
  return div;
}
function singletonRef(elem) {
  if (elem) {
    globalToast = elem;
  }
}

/**
 * 用于多次重复 renderGlobalToast 和销毁自动生成的 DOM 元素
 */
class ToastWrapper extends React.Component {
  constructor() {
    super(...arguments);
    _defineProperty(this, "div", void 0);
    _defineProperty(this, "isDefaultContainer", true);
    _defineProperty(this, "getContainer", () => {
      if (!this.div) {
        var _Toast$globalConfig$g, _Toast$globalConfig;
        const globalContainer = (_Toast$globalConfig$g = (_Toast$globalConfig = Toast.globalConfig).getContainer) === null || _Toast$globalConfig$g === void 0 ? void 0 : _Toast$globalConfig$g.call(_Toast$globalConfig);
        this.div = globalContainer || getDefaultContainer();
        if (globalContainer) {
          this.isDefaultContainer = false;
        }
      }
      return this.div;
    });
  }
  componentWillUnmount() {
    globalToast = undefined;
    if (this.isDefaultContainer && this.div) {
      this.div.remove();
      this.div = undefined;
    }
  }
  render() {
    return /*#__PURE__*/React.createElement(Toast, _extends({}, this.props, {
      ref: singletonRef,
      getContainer: this.getContainer
    }));
  }
}
export function getGlobalToast() {
  if (globalToast) {
    return Promise.resolve(globalToast);
  }
  if (globalToastInitPromise) {
    return globalToastInitPromise;
  }
  globalToastInitPromise = new Promise(resolve => {
    const div = getDefaultContainer();
    // 如果没有使用 useToast 或者 renderGlobalToast 初始化 toast，
    // 则默认取第一个 ConfigProvider 提供的配置
    ReactDOM.render(/*#__PURE__*/React.createElement(CrossRootConfigProvider, null, /*#__PURE__*/React.createElement(ToastWrapper, {
      ref: () => {
        if (globalToast) {
          globalToastInitPromise = undefined;
          resolve(globalToast);
        }
      },
      getContainer: () => div
    })), div);
  });
  return globalToastInitPromise;
}
function createTostWrapper(getToast) {
  return {
    open(config) {
      return getToast().then(toast => toast.open(config) || '');
    },
    info(configOrContent, duration, onClose) {
      return getToast().then(toast => toast.info(configOrContent, duration, onClose));
    },
    success(configOrContent, duration, onClose) {
      return getToast().then(toast => toast.success(configOrContent, duration, onClose));
    },
    warning(configOrContent, duration, onClose) {
      return getToast().then(toast => toast.warning(configOrContent, duration, onClose));
    },
    warn(configOrContent, duration, onClose) {
      return getToast().then(toast => toast.warn(configOrContent, duration, onClose));
    },
    error(configOrContent, duration, onClose) {
      return getToast().then(toast => toast.error(configOrContent, duration, onClose));
    },
    loading(configOrContent, duration, onClose) {
      return getToast().then(toast => toast.loading(configOrContent, duration, onClose));
    },
    config(options) {
      Toast.globalConfig = Object.assign(Toast.globalConfig, options);
    },
    destroy(key) {
      getToast().then(toast => toast.destroy(key));
    }
  };
}
export { Toast };
export const toast = createTostWrapper(getGlobalToast);

/**
 * 用于创建可以获取 ConfigContext 的全局 Toast 单例
 * @example
 * ```tsx
 * <ConfigProvider>
 *   {renderGlobalToast({ className: 'my-toast' }, document.getElementById('my-container'))}
 *   <App />
 * </ConfigProvider>
 * ```
 *
 * @param props: ToastProps
 * @returns React.ReactPortal
 */
export function renderGlobalToast(props) {
  return /*#__PURE__*/React.createElement(ToastWrapper, props);
}

/**
 * 获取一个局部的 toast 实例
 * @returns
 */
export function useToast(props, container) {
  const toastRef = useRef(null);
  const mountDivRef = useRef();
  const toastIns = createTostWrapper(() => {
    if (toastRef.current) {
      return Promise.resolve(toastRef.current);
    }
    return getGlobalToast();
  });
  const _getDefaultContainer = () => {
    if (mountDivRef.current) {
      return mountDivRef.current;
    }
    mountDivRef.current = getDefaultContainer();
    return mountDivRef.current;
  };
  return [toastIns, /*#__PURE__*/React.createElement(ConfigProvider, {
    key: "ud-hook-toast-wrapper"
  }, /*#__PURE__*/React.createElement(Toast, _extends({
    key: "ud-hook-toast"
  }, props, {
    ref: toastRef,
    getContainer: () => container || _getDefaultContainer()
  })))];
}

// 兼容竞品 API
export const message = toast;
export const useMessage = useToast;
