// @universe-design/react/es/message/Toast.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _extends from "@babel/runtime/helpers/esm/extends";
import React, { createRef } from 'react';
import clsx from 'clsx';
import omit from 'lodash/omit';
import pick from 'lodash/pick';
import stylex from '@universe-design/stylex';
import { cssTypographyHeadline } from '../stylex';
import { Notice } from '../notice';
import NotificationManager, {
// BASE_MESSAGE_LENGTH,
isNotAutoClose } from '../notification/manager';
import './style';
const ManagerProps = ['className', 'prefixCls', 'duration', 'transitionProps', 'maxCount', 'getContainer', 'top'];

/**
 * `toast.config` 接收的参数。
 * @public
 * @docgen
 */

/**
 * toast 内容配置。
 * @public
 * @docgen
 */

function getActionDuration(action, duration) {
  // eslint-disable-next-line @typescript-eslint/no-magic-numbers
  const actionDur = action ? 6000 : undefined;
  return duration !== undefined ? duration : actionDur;
}
function getManagerConfig(option) {
  const {
    content,
    key,
    duration,
    onClose,
    className,
    type,
    showCloseIcon,
    newStyle = true,
    ...rest
  } = option;
  // let contentKey: string | undefined;
  // let contentLength = 0;
  // if (typeof content === 'string') {
  //   contentKey = `${type}-${content}`;
  //   contentLength = content.length;
  // }
  // // 显示关闭按钮的逻辑
  // // 1. 设置不自动关闭时
  // // 2. 全局配置 closable 为 true
  // // 3. 内容长度 >= BASE_MESSAGE_LENGTH
  // // 4. 内容是 object 类型时，包含 null
  // const newClosable =
  //   isNotAutoClose(duration) ||
  //   showCloseIcon ||
  //   contentLength >= BASE_MESSAGE_LENGTH ||
  //   typeof content === 'object';
  const newClosable = showCloseIcon !== undefined ? showCloseIcon : isNotAutoClose(duration);
  return {
    key,
    duration: getActionDuration(rest.action, duration),
    renderMessage: _ref => {
      let {
        close,
        rootPrefixCls,
        toastConfigCls
      } = _ref;
      return /*#__PURE__*/React.createElement(Notice, _extends({}, rest, {
        closable: newClosable,
        visible: true,
        className: clsx(`${rootPrefixCls}__toast`, className, toastConfigCls || ''),
        theme: newStyle ? 'outlined' : 'plain',
        message: /*#__PURE__*/React.createElement("div", {
          className: stylex(cssTypographyHeadline)
        }, content),
        type: type,
        onClose: close,
        banner: false
        // maxLine={2}
      }));
    },
    onClose
  };
}

/**
 * 1. 不需要 theme 参数，只有一种样式
 * 2. info 使用目前的 normal 样式
 * 3. 其它使用目前的 light 模式
 * 4. export Toast 可挂载到当前容器
 * 5. export message、toast 直接调起 API
 */
export class Toast extends React.PureComponent {
  constructor() {
    super(...arguments);
    _defineProperty(this, "managerRef", /*#__PURE__*/createRef());
  }
  /**
   * 计算后的配置
   */
  get _config() {
    return Object.assign({}, Toast.globalConfig, this.props);
  }

  /* eslint-disable no-dupe-class-members */
  /**
   * 返回一个 onClose 执行时 resolve 的 promise
   * @param config
   * @returns
   */
  open(config) {
    return new Promise(resolve => {
      var _this$managerRef$curr;
      const handleClose = config.onClose;
      config.onClose = () => {
        handleClose && handleClose();
        resolve(undefined);
      };
      (_this$managerRef$curr = this.managerRef.current) === null || _this$managerRef$curr === void 0 ? void 0 : _this$managerRef$curr.add(this._getManagerConfig(config));
    });
  }
  info(configOrContent, duration, onClose) {
    return this._open('info', configOrContent, duration, onClose);
  }
  success(configOrContent, duration, onClose) {
    return this._open('success', configOrContent, duration, onClose);
  }
  warning(configOrContent, duration, onClose) {
    return this._open('warning', configOrContent, duration, onClose);
  }
  warn(configOrContent, duration, onClose) {
    return this.warning(configOrContent, duration, onClose);
  }
  error(configOrContent, duration, onClose) {
    return this._open('error', configOrContent, duration, onClose);
  }
  loading(configOrContent, duration, onClose) {
    return this._open('loading', configOrContent, duration, onClose);
  }
  config(props) {
    // TODO 这里没有想好，之前是 assign 但是感觉直接覆盖 config 比较好，要不有些值写入了就没法删掉了
    Toast.globalConfig = Object.assign(Toast.globalConfig, props);
    // Toast.config = props;
    this.forceUpdate();
  }
  destroy(key) {
    if (key) {
      var _this$managerRef$curr2;
      (_this$managerRef$curr2 = this.managerRef.current) === null || _this$managerRef$curr2 === void 0 ? void 0 : _this$managerRef$curr2.remove(key);
    } else {
      var _this$managerRef$curr3;
      (_this$managerRef$curr3 = this.managerRef.current) === null || _this$managerRef$curr3 === void 0 ? void 0 : _this$managerRef$curr3.destroy();
    }
  }
  render() {
    // Debug
    // console.log('Toast render', { config: this._config });
    return /*#__PURE__*/React.createElement(NotificationManager, _extends({
      ref: this.managerRef,
      placement: "top",
      transitionProps: {
        timeout: 300,
        classNames: 'toast--animation'
      },
      duration: 3000
    }, pick(this._config, ManagerProps)));
  }
  _getManagerConfig(config) {
    const props = Object.assign({}, omit(this._config, ManagerProps));
    return getManagerConfig({
      ...props,
      ...config
    });
  }
  _open(type, configOrContent, duration, onClose) {
    if (configOrContent && typeof configOrContent === 'object' && 'content' in configOrContent) {
      return this.open({
        ...configOrContent,
        type
      });
    } else {
      const d = typeof duration === 'function' ? duration() : duration;
      return this.open({
        type,
        content: configOrContent,
        duration: d,
        onClose
      });
    }
  }
  /* eslint-enable no-dupe-class-members */
}
/**
 * 全局 config
 */
_defineProperty(Toast, "globalConfig", {});
