// @universe-design/react/src/drawer/stories/homeless.stories.tsx
import type { StoryFn } from '@storybook/react';
import React, { useState } from 'react';
import type { DrawerProps } from '..';
import { Drawer } from '..';
import type { ButtonProps } from '../..';
import { Button, Input, Dialog } from '../..';

export default {
  title: 'Components/Drawer/临时示例',
  component: Drawer,
};

const RButton = (props: ButtonProps) => <Button style={{ marginRight: 16 }} {...props} />;

export const DifferentStates = () => {
  const [state, setState] = useState<DrawerProps>({
    visible: false,
    size: 'large',
  });

  const onClose = () => {
    setState(prevProps => ({
      ...prevProps,
      visible: false,
    }));
  };

  const showDrawerByProps = (drawerProps: DrawerProps) => () => {
    setState({
      visible: true,
      ...drawerProps,
    });
  };

  const showDrawerByPlacement = (placement: DrawerProps['placement']) => () => {
    setState({
      placement,
      visible: true,
    });
  };

  const showDrawerByContainer = (getContainer: DrawerProps['getContainer']) => () => {
    setState({
      getContainer,
      visible: true,
    });
  };
  const sizeTest: { [_: string]: DrawerProps } = {
    small: { size: 'small' },
    medium: { size: 'medium' },
    large: { size: 'large' },
    'large-top': {
      size: 'large',
      placement: 'top',
    },
    'large-bottom': {
      size: 'large',
      placement: 'bottom',
    },
    'large-width(200px)': {
      size: 'large',
      width: '200px',
    },
  };

  return (
    <div>
      <div>
        <label>尺寸：</label>
        <p>仅对 placement 为 right、left 时生效</p>
        {Object.entries(sizeTest).map(([key, props]) => (
          <RButton key={key} type="primary" onClick={showDrawerByProps(props)}>
            {key}
          </RButton>
        ))}
      </div>
      <div style={{ marginTop: 20 }}>
        <label>方向：</label>
        <RButton type="primary" onClick={showDrawerByPlacement('top')}>
          top
        </RButton>
        <RButton type="primary" onClick={showDrawerByPlacement('right')}>
          right
        </RButton>
        <RButton type="primary" onClick={showDrawerByPlacement('bottom')}>
          bottom
        </RButton>
        <RButton type="primary" onClick={showDrawerByPlacement('left')}>
          left
        </RButton>
      </div>
      <div style={{ marginTop: 20 }}>
        <label>位置：</label>
        <RButton type="primary" onClick={showDrawerByContainer(undefined)}>
          对话框-body
        </RButton>
        <RButton type="primary" onClick={showDrawerByContainer(false)}>
          对话框-文档流中
        </RButton>
        <RButton
          type="primary"
          onClick={showDrawerByProps({
            getContainer: false,
            mask: false,
            placement: 'bottom',
          })}
        >
          文档流中
        </RButton>
      </div>
      <div style={{ marginTop: 20 }}>
        <label>bodyPaddingSize: </label>
        <RButton type="primary" onClick={showDrawerByProps({})}>
          默认（24）
        </RButton>
        <RButton type="primary" onClick={showDrawerByProps({ bodyPaddingSize: 'small' })}>
          small（20）
        </RButton>
        <RButton type="primary" onClick={showDrawerByProps({ bodyPaddingSize: 48 })}>
          自定义（48）
        </RButton>
      </div>
      <div style={{ marginTop: 20 }}>
        <label>可定制区域：</label>
        <RButton
          type="primary"
          onClick={showDrawerByProps({
            showCloseIcon: true,
            title: (
              <div
                style={{
                  display: 'flex',
                  justifyContent: 'space-between',
                  alignItems: 'center',
                }}
              >
                <span>标题</span>
                <span>
                  <span style={{ fontWeight: 'normal' }}>定制操作</span>
                  <span
                    style={{
                      margin: '0 20px 0 20px',
                      lineHeight: '20px',
                      borderLeft: '1px solid #BBBFC4',
                      // verticalAlign: 'middle',
                    }}
                  />
                </span>
              </div>
            ),
          })}
        >
          header 操作
        </RButton>
        <RButton type="primary" onClick={showDrawerByProps({ subTitle: '副标题' })}>
          副标题
        </RButton>
        <RButton
          type="primary"
          onClick={showDrawerByProps({
            footer: (
              <div>
                <Button type="primary" style={{ marginRight: 12 }}>
                  主要按钮
                </Button>
                <Button>次要按钮</Button>
              </div>
            ),
          })}
        >
          footer
        </RButton>
        <RButton
          type="primary"
          onClick={showDrawerByProps({
            headerExtra: <div>headerExtra</div>,
          })}
        >
          headerExtra
        </RButton>
      </div>
      <div style={{ marginTop: 20 }}>
        <label>其他：</label>
        <RButton type="primary" onClick={showDrawerByProps({ showCloseIcon: true })}>
          显示关闭按钮
        </RButton>
      </div>
      <Input />
      <div style={{ marginTop: 20 }}>
        <Drawer title="标题" onClose={onClose} {...state}>
          <p>Some contents...</p>
          <p>Some contents...</p>
          <p>Some contents...</p>
          <Input />
          <Input />
          <Input />
          <Button>确定</Button>
        </Drawer>
      </div>
    </div>
  );
};
DifferentStates.storyName = 'z. 各种不同的状态';

export const ScrollContent = () => (
  <Drawer
    title="标题"
    visible
    bodyPaddingSize={0}
    width={400}
    footer={
      <div style={{ textAlign: 'right' }}>
        <Button style={{ marginRight: 12 }}>次要按钮</Button>
        <Button type="primary">主要按钮</Button>
      </div>
    }
  >
    <div style={{ height: '200vh' }}>Some contents...</div>
    <div>Some contents...</div>
    <div>Some contents...</div>
  </Drawer>
);
ScrollContent.storyName = 'z. 内容滚动';

export const RenderToCustomContainer = () => (
  <div id="test">
    <Drawer
      title="Basic Drawer"
      visible
      getContainer={() => {
        const container = document.getElementById('test');
        console.log('这里的 container 是: ', container);
        return container;
      }}
    >
      <p>Some contents...</p>
      <p>Some contents...</p>
      <p>Some contents...</p>
    </Drawer>
  </div>
);
RenderToCustomContainer.storyName = 'z. 渲染到自定义容器';

export const DialogInDrawer = () => {
  const [visible, setVisible] = React.useState(false);
  return (
    <>
      <Button onClick={() => setVisible(true)}>Open Drawer</Button>
      <Drawer title="Basic drawer" visible={visible} onClose={() => setVisible(false)}>
        <Button>Fake button</Button>
        <Button
          onClick={() =>
            Dialog.confirm({
              title: 'Basic confirm',
              // content: <Input />,
              // autoFocusButton: null,
              // cancelButtonProps: {
              //   autoFocus: true,
              // }
            })
          }
        >
          Open confirm
        </Button>
      </Drawer>
    </>
  );
};
DialogInDrawer.storyName = 'z. 抽屉里的对话框';

export const MultiLevelAndChangeWidth: StoryFn = () => {
  const [visible, setVisible] = useState(false);
  const [visible2, setVisible2] = useState(false);

  const [v, setV] = useState(400);

  const [width, setWidth] = useState(400);

  console.log({
    width,
    v,
    visible,
    visible2,
  });
  return (
    <>
      <Button
        type="primary"
        onClick={e => {
          setVisible(true);
        }}
      >
        Open
      </Button>
      <Drawer
        title="Basic Drawer 1"
        visible={visible}
        onClose={e => {
          console.log(e);
          setVisible(false);
        }}
        width={visible2 ? 480 : 400}
      >
        <p>1111111111111</p>
        <Button
          type="primary"
          onClick={e => {
            setVisible2(true);
          }}
        >
          Open22222222
        </Button>

        <Drawer
          title="Basic Drawer 2"
          visible={visible2}
          onClose={e => {
            console.log(e);
            setVisible2(false);
          }}
          width={width}
        >
          <Button
            type="primary"
            onClick={e => {
              setVisible2(false);
            }}
          >
            Close22222222
          </Button>

          <Input value={String(v)} onChange={e => setV(Number(e.target.value) as any)} />

          <Button
            type="primary"
            onClick={e => {
              setWidth(v);
            }}
          >
            变更宽度为{v}
          </Button>
          <p>222222222</p>
        </Drawer>
      </Drawer>
    </>
  );
};

MultiLevelAndChangeWidth.storyName = '多级抽屉变更宽度';
