// @universe-design/react/src/drawer/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import CBasic from '../demo/basic';
import CCustomPlacement from '../demo/customPlacement';
import CForm from '../demo/form';
import CRenderInDom from '../demo/renderInDom';
if (!(CBasic as Story).storyName) {
  (CBasic as Story).storyName = '0. 基础用法';
  (CBasic as Story).parameters = {
    docs: {
      description: {
        story: '简单的用法。点击触发按钮抽屉从右滑出，点击遮罩区关闭。\n\n',
      },
    },
  };
}
if (!(CCustomPlacement as Story).storyName) {
  (CCustomPlacement as Story).storyName = '1. 自定义位置';
  (CCustomPlacement as Story).parameters = {
    docs: {
      description: {
        story: '自定义位置，点击触发按钮抽屉从相应的位置滑出，点击遮罩区关闭。\n\n',
      },
    },
  };
}
if (!(CForm as Story).storyName) {
  (CForm as Story).storyName = '3. 抽屉表单';
  (CForm as Story).parameters = {
    docs: {
      description: {
        story: '在抽屉中使用表单。\n\n',
      },
    },
  };
}
if (!(CRenderInDom as Story).storyName) {
  (CRenderInDom as Story).storyName = '2. 挂载节点';
  (CRenderInDom as Story).parameters = {
    docs: {
      description: {
        story: '可以通过 `getPopupContainer` 指定抽屉挂载的父级节点。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Drawer/官网示例',
  tags: ['autodocs'],
};

export { CBasic, CCustomPlacement, CForm, CRenderInDom };
