// @universe-design/react/src/upload/uploadList/index.tsx
import React from 'react';
import clsx from 'clsx';
import { UploadCardFile, UploadPictureFile, UploadSimpleTextFile, UploadTextFile } from '../uploadFile';
import { useUploadContext, useUploadContextSelector } from '../hooks/UploadContext';
import { Popover } from '../../popover';
import { Space } from '../../space';
import type { UploadListProps } from './interface';
import type { UploadFile } from '../uploadFile/interface';

// UploadCardFileList
export interface UploadCardFileListProps extends UploadListProps {
  /**
   * 自定义列表项, 可用来绑定事件回调，自定义 icon 等
   * @defaultValue (file: UploadFile) => <UploadCardFile file={file} />
   */
  itemRenderer?: UploadListProps['itemRenderer'];
}
const defaultCardItemRenderer = (file: UploadFile) => <UploadCardFile file={file} />;

export const UploadCardFileList = (props: UploadListProps) => {
  const { itemRenderer = defaultCardItemRenderer, ...spaceProps } = props;

  const fileList = useUploadContextSelector(ctx => ctx.fileList);
  return (
    <Space
      inline={false}
      direction="horizontal"
      // eslint-disable-next-line @typescript-eslint/no-magic-numbers
      size={[40, 16]}
      wrap
      align="start"
      {...spaceProps}
      itemKey={index => fileList[index].uid || index}
    >
      {fileList.map(file => (
        <React.Fragment key={file.uid}>{itemRenderer(file)}</React.Fragment>
      ))}
    </Space>
  );
};

// UploadTextFileList
export interface UploadTextFileListProps extends UploadListProps {
  /**
   * 自定义列表项, 可用来绑定事件回调，自定义 icon 等
   * @defaultValue (file: UploadFile) => <UploadTextFile file={file} />
   */
  itemRenderer?: UploadListProps['itemRenderer'];
}
const defaultTextItemRenderer = (file: UploadFile) => <UploadTextFile file={file} />;

export const UploadTextFileList = (props: UploadTextFileListProps) => {
  const { itemRenderer = defaultTextItemRenderer, ...spaceProps } = props;

  const fileList = useUploadContextSelector(ctx => ctx.fileList);
  return (
    <Space direction="vertical" inline={false} {...spaceProps} itemKey={index => fileList[index].uid || index}>
      {fileList.map(file => (
        <React.Fragment key={file.uid}>{itemRenderer(file)}</React.Fragment>
      ))}
    </Space>
  );
};

// UploadSimpleTextFileList
export interface UploadSimpleTextFileListProps extends UploadListProps {
  /**
   * 自定义列表项, 可用来绑定事件回调，自定义 icon 等
   * @defaultValue (file: UploadFile) => <UploadSimpleTextFile file={file} />
   */
  itemRenderer?: UploadListProps['itemRenderer'];
}
const defaultSimpleTextItemRenderer = (file: UploadFile) => <UploadSimpleTextFile file={file} />;
export const UploadSimpleTextFileList = (props: UploadSimpleTextFileListProps) => {
  const { itemRenderer = defaultSimpleTextItemRenderer, ...spaceProps } = props;

  const fileList = useUploadContextSelector(ctx => ctx.fileList);
  return (
    <Space direction="vertical" inline={false} {...spaceProps} itemKey={index => fileList[index].uid || index}>
      {fileList.map(file => (
        <React.Fragment key={file.uid}>{itemRenderer(file)}</React.Fragment>
      ))}
    </Space>
  );
};

// UploadPictureFileList
/**
 * @public
 */
export interface UploadPictureFileListProps extends UploadListProps {
  /**
   * 最多展示几张图片，多余图片将被收起在 `Popover` 里面（listType为picture才生效）
   */
  maxDisplayedPictureCount?: number;
  /**
   * 图片墙上传按钮
   */
  uploadTrigger?: React.ReactNode;
  /**
   * 自定义列表项, 可用来绑定事件回调，自定义 icon 等
   * @defaultValue (file: UploadFile) => <UploadPictureFile file={file} />
   */
  itemRenderer?: UploadListProps['itemRenderer'];
}

export const UploadPictureFileList = (props: UploadPictureFileListProps) => {
  const {
    maxDisplayedPictureCount,
    uploadTrigger,
    itemRenderer = (file: UploadFile) => <UploadPictureFile file={file} />,
    ...spaceProps
  } = props;

  const { fileList, rootPrefixCls } = useUploadContext();
  let extraPictureList: UploadFile[] = [];
  let defaultDisplayedFiles = fileList;

  if (maxDisplayedPictureCount && fileList.length > maxDisplayedPictureCount) {
    extraPictureList = fileList.slice(maxDisplayedPictureCount - 1);
    defaultDisplayedFiles = fileList.slice(0, maxDisplayedPictureCount - 1);
  }

  const firstExtraFile = extraPictureList[0];
  const imgSrc = firstExtraFile ? firstExtraFile.thumbUrl || firstExtraFile.url : null;

  /**
   * space 默认会使用 index 作为 itemKey
   * 会使得 uploadTrigger 的 index 不断增加。导致 key 变化引发节点销毁，
   * 进而上传动作被取消
   */
  const itemKey = (index: number) => {
    if (index < defaultDisplayedFiles.length) {
      return String(defaultDisplayedFiles[index].uid);
    }
    if (extraPictureList.length && index === defaultDisplayedFiles.length) {
      return String(extraPictureList[0].uid);
    }
    return 'upload-trigger';
  };

  return (
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    <Space wrap inline={false} size={[12, 6]} itemKey={itemKey} {...spaceProps}>
      {defaultDisplayedFiles.map(file => (
        <React.Fragment key={file.uid}>{itemRenderer(file)}</React.Fragment>
      ))}
      {extraPictureList.length > 0 && (
        <Popover
          destroyOnHide={false}
          placement="bottom"
          content={
            <div className={`${rootPrefixCls}-file-list ${rootPrefixCls}-item--picture__extra__list`}>
              {extraPictureList.map(_file => (
                <React.Fragment key={_file.uid}>{itemRenderer(_file)}</React.Fragment>
              ))}
            </div>
          }
        >
          <div
            className={clsx(
              `${rootPrefixCls}-item`,
              `${rootPrefixCls}-item--picture`,
              `${rootPrefixCls}-item--picture__extra`,
            )}
          >
            {imgSrc && <img src={imgSrc} />}
            <div className={`${rootPrefixCls}-item--picture-mask`} />
            <div className={`${rootPrefixCls}-item--picture__extra-count`} data-testid="packedCount">
              +{extraPictureList.length}
            </div>
          </div>
        </Popover>
      )}
      {uploadTrigger}
    </Space>
  );
};
