// @universe-design/react/src/shared/components/AutoSizer.tsx
/**
 * Forked and edited from https://github.com/bvaughn/react-virtualized-auto-sizer/blob/master/src/index.js
 * Thanks for bvaughn's work!
 */
import React, { useRef, useState, useMemo } from 'react';
import type { ForwardedRef } from 'react';
import { createResizeObserver } from '@universe-design/react-utilities';
import { useEventCallback, useIsomorphicLayoutEffect as useLayoutEffect } from '../../hooks';
import composeRefs from '../../util/composeRefs';

type Size = {
  height: number;
  width: number;
};

export interface AutoSizerProps {
  /**
   * 渲染函数
   */
  children: (size: Size) => React.ReactElement;
  /**
   * `className` 属性透传
   */
  className?: string;
  /**
   * `style` 属性透传
   */
  style?: React.CSSProperties;
  /**
   * 默认高度，SSR 时有效
   * @defaultValue 0
   */
  defaultHeight?: number;
  /**
   * 默认宽度，SSR 时有效
   * @defaultValue 0
   */
  defaultWidth?: number;
  /**
   * 是否禁用动态高度
   * @defaultValue false
   */
  disableHeight?: boolean;
  /**
   * 是否禁用动态宽度
   * @defaultValue false
   */
  disableWidth?: boolean;
  /**
   * 尺寸变化回调函数
   */
  onResize?: (size: Size) => void;
}

export const AutoSizer = React.forwardRef((props: AutoSizerProps, ref: ForwardedRef<HTMLDivElement>) => {
  /** Props */
  const {
    defaultWidth = 0,
    defaultHeight = 0,
    disableWidth = false,
    disableHeight = false,
    className,
    style,
    onResize,
    children,
  } = props;

  /** State */
  const [width, setWidth] = useState(defaultWidth);
  const [height, setHeight] = useState(defaultHeight);

  /** Refs */
  const autoSizerRef = useRef<HTMLDivElement>(null);
  const parentNodeRef = useRef<HTMLElement>();

  const handleResize = useEventCallback<ResizeObserverCallback>(entries => {
    const parentNode = parentNodeRef.current;
    for (const entry of entries) {
      if (entry.target !== parentNode || !parentNode) {
        continue;
      }

      const contentBoxSize: ResizeObserverSize | undefined = Array.isArray(entry.contentBoxSize)
        ? entry.contentBoxSize[0]
        : entry.contentBoxSize; // firefox 不是数组，safari < 15.4 没有 contentBoxSize

      const size: Size = (() => {
        if (contentBoxSize) {
          return {
            width: contentBoxSize.inlineSize,
            height: contentBoxSize.blockSize,
          };
        }

        const _width = parentNode.offsetWidth;
        const _height = parentNode.offsetHeight;
        const cstyle = getComputedStyle(parentNode);
        // TODO 暂时只支持 px
        const paddingLeft = parseInt(cstyle.paddingLeft, 10) || 0;
        const paddingRight = parseInt(cstyle.paddingRight, 10) || 0;
        const paddingTop = parseInt(cstyle.paddingTop, 10) || 0;
        const paddingBottom = parseInt(cstyle.paddingBottom, 10) || 0;

        return {
          width: _width - paddingLeft - paddingRight,
          height: _height - paddingTop - paddingBottom,
        };
      })();

      const widthChanged = !disableWidth && width !== size.width;
      const heightChanged = !disableHeight && height !== size.height;

      if (widthChanged) {
        setWidth(size.width);
      }

      if (heightChanged) {
        setHeight(size.height);
      }

      if (widthChanged || heightChanged) {
        onResize?.(size);
      }
    }
  });

  useLayoutEffect(() => {
    const autoSizer = autoSizerRef.current;
    const parentNode = autoSizer?.parentNode;
    const HTMLElement = parentNode?.ownerDocument?.defaultView?.HTMLElement;
    if (!parentNode || !HTMLElement || !(parentNode instanceof HTMLElement)) {
      return;
    }
    parentNodeRef.current = parentNode;
    const ro = createResizeObserver();
    ro.subscribe(handleResize);
    ro.observer.observe(parentNode, { box: 'content-box' });
    return () => {
      ro.unsubscribe(handleResize);
      ro.observer.disconnect();
    };
  }, [handleResize]);

  return (
    <div
      className={className}
      style={{
        overflow: 'visible',
        ...style,
      }}
      ref={useMemo(() => composeRefs(ref, autoSizerRef), [ref])}
    >
      {children({ width, height })}
    </div>
  );
});
