// @universe-design/react/src/shared/components/FirstLineAlign.tsx
/** @jsxRuntime classic */
/** @jsx jsx */
import stylex from '@universe-design/stylex';
import clsx from 'clsx';
import * as React from 'react';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx, useMergeComponentProps } from '@universe-design/stylex/react';
import { cssFlexBase } from '../../stylex';
import { forwardHTMLRef } from '../../util/forwardRef';
import type { StylexProps } from '../../util/types';
import { useCssScope } from '../../config-provider';

/**
 * 用于撑开 flex 父级
 */
export const EmptyElement = forwardHTMLRef<StylexProps, 'span'>((props, ref) => (
  <span
    ref={ref}
    {...useMergeComponentProps(props, {
      className: useCssScope('empty-inline-element')(''),
      sx: [
        stylex.create({
          width: 0,
          overflow: 'hidden',
          display: 'inline-block', // 不能使用 inline 因为 inline 不能调整宽度
        }),
      ],
    })}
  >
    {'\xa0'}
  </span>
));

if (__DEV__) {
  EmptyElement.displayName = 'EmptyElement';
}

/**
 * 用于首行对齐
 */
export const FirstLineAlign = React.forwardRef<
  React.ComponentRef<'div'>,
  React.ComponentPropsWithoutRef<'div'> &
    StylexProps & {
      beforeProps?: React.ComponentPropsWithoutRef<'div'> & StylexProps;
    }
>(({ beforeProps, children, className, sx = [], ...restProps }, ref) => {
  const before = (() => {
    if (!beforeProps) {
      return null;
    }
    const { className: beforeClassName, children: beforeChildren, sx: beforeSx = [], ...beforeRestProps } = beforeProps;
    return (
      <div
        className={clsx(
          beforeClassName,
          stylex(
            cssFlexBase,
            stylex.create({
              alignItems: 'baseline',
              flexGrow: 0,
              flexShrink: 0,
              flexBasis: 'auto',
            }),
            ...beforeSx,
          ),
        )}
        {...beforeRestProps}
      >
        {beforeChildren}
        <EmptyElement />
      </div>
    );
  })();
  return (
    <div
      className={clsx(
        className,
        stylex(
          cssFlexBase,
          stylex.create({
            alignItems: 'baseline',
          }),
          ...sx,
        ),
      )}
      ref={ref}
      {...restProps}
    >
      {before}
      {children}
      <EmptyElement />
    </div>
  );
});
