// @universe-design/react/src/shared/components/Image.tsx
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx, makeStylexStyles } from '@universe-design/stylex/react';
import * as React from 'react';
import type { StylexProps } from '@universe-design/stylex/react';
import stylex from '@universe-design/stylex';
import { useEventCallback } from '@universe-design/react-utilities';
import { composeEventHandlers } from '../../util/function';
import { forwardHTMLRef } from '../../internal';
import { useCssScope } from '../../config-provider/hooks';

const loading = 'loading';
const error = 'error';
const loaded = 'loaded';

type ImageLoadStatus = typeof loading | typeof error | typeof loaded;

const useImageStylexStyles = makeStylexStyles(({ prefix }: { prefix: string }) => ({
  base: stylex.create(
    {
      boxSizing: 'border-box',
      display: 'inline-block',
    },
    prefix,
  ),
  status: {
    [loading]: stylex.create({}, `${prefix}--loading`),
    [error]: stylex.create({}, `${prefix}--error`),
    [loaded]: stylex.create({}, `${prefix}--loaded`),
  },
  // 图片适配尺寸策略
  fit: {
    center: stylex.create(
      {
        objectFit: 'none',
        objectPosition: 'center',
        height: '100%',
        width: '100%',
      },
      `${prefix}--fit-center`,
    ),
    contain: stylex.create(
      {
        objectFit: 'contain',
        objectPosition: 'center',
        height: '100%',
        width: '100%',
      },
      `${prefix}--fit-contain`,
    ),
    default: stylex.create({}),
    cover: stylex.create(
      {
        objectFit: 'cover',
        objectPosition: 'center',
        height: '100%',
        width: '100%',
      },
      `${prefix}--fit-cover`,
    ),
    none: stylex.create(
      {
        objectFit: 'none',
        objectPosition: 'left top',
        height: '100%',
        width: '100%',
      },
      `${prefix}--fit-none`,
    ),
  },
}));

export const useImage_unstable = (
  props: {
    fit?: 'none' | 'center' | 'contain' | 'cover' | 'default';
  } & React.ComponentPropsWithoutRef<'img'> &
    StylexProps,
  ref: React.Ref<HTMLImageElement>,
) => {
  const { fit = 'default', sx, ...htmlProps } = props;

  const styles = useImageStylexStyles({ prefix: useCssScope('image')() });

  const [status, setStatus] = React.useState<ImageLoadStatus>(loading);

  React.useMemo(() => {
    props.src;
    setStatus(loading);
  }, [props.src]);

  const imageProps = {
    ...htmlProps,
    ref,
    sx: [styles.base, styles.fit[fit], styles.status[status], sx],
    onLoad: useEventCallback(
      composeEventHandlers(props.onLoad, () => {
        setStatus(loaded);
      }),
    ),
    onError: useEventCallback(
      composeEventHandlers(props.onError, () => {
        setStatus(error);
      }),
    ),
  } satisfies React.ComponentProps<'img'> & StylexProps;

  const state = {
    status,
  };

  return {
    state,
    slotsProps: {
      root: imageProps,
    },
  };
};

export type UseImageProps = Parameters<typeof useImage_unstable>[0];

export type ImageProps = UseImageProps;

export const Image_unstable = forwardHTMLRef<ImageProps, 'img'>((props, ref) => {
  const { slotsProps } = useImage_unstable(props, ref);
  return <img {...slotsProps.root} />;
});
