// @universe-design/react/src/expandable-text/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Basic from '../demo/0.basic';
import C1Typography from '../demo/1.typography';
import C2TriggerPosition from '../demo/2.trigger-position';
import C3Background from '../demo/3.background';
import C4LineClamp from '../demo/4.line-clamp';
import C5PreventCollapse from '../demo/5.prevent-collapse';
if (!(C0Basic as Story).storyName) {
  (C0Basic as Story).storyName = '0. 基础用法';
  (C0Basic as Story).parameters = {
    docs: {
      description: {
        story: '简单的用法。\n\n',
      },
    },
  };
}
if (!(C1Typography as Story).storyName) {
  (C1Typography as Story).storyName = '1. 不同排版样式';
  (C1Typography as Story).parameters = {
    docs: {
      description: {
        story:
          '可展开文本包括「段落」和「触发器」两个组件，这两个组件都会继承父级的排版样式，下面是使用正文辅助排版的示例。\n\n',
      },
    },
  };
}
if (!(C2TriggerPosition as Story).storyName) {
  (C2TriggerPosition as Story).storyName = '2. 触发器位置';
  (C2TriggerPosition as Story).parameters = {
    docs: {
      description: {
        story: '触发器默认处于行尾，可以同通过 `triggerProps.position` 设置不同的触发器位置。\n\n',
      },
    },
  };
}
if (!(C3Background as Story).storyName) {
  (C3Background as Story).storyName = '3. 带背景的段落';
  (C3Background as Story).parameters = {
    docs: {
      description: {
        story:
          '当段落出现在有色背景上时，可以通过 `triggerProps.maskColor` 设置**行尾布局并收起时**的触发器蒙层颜色为有色背景颜色，从而保证颜色连贯。\n\n',
      },
    },
  };
}
if (!(C4LineClamp as Story).storyName) {
  (C4LineClamp as Story).storyName = '4. 截断行数';
  (C4LineClamp as Story).parameters = {
    docs: {
      description: {
        story:
          '默认在 3 行处截断，可以通过 `lineClamp` 设置自定义截断行数，如果 `lineClamp` 小于 1 则会当做 1 处理。\n\n',
      },
    },
  };
}
if (!(C5PreventCollapse as Story).storyName) {
  (C5PreventCollapse as Story).storyName = '5. 一次性展开';
  (C5PreventCollapse as Story).parameters = {
    docs: {
      description: {
        story:
          '默认情况下段落可以反复展开和收起，可以通过 `preventCollapseWhenExpanded` 设置为一次性展开，不能收起。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/ExpandableText/官网示例',
  tags: ['autodocs'],
};

export { C0Basic, C1Typography, C2TriggerPosition, C3Background, C4LineClamp, C5PreventCollapse };
