// @universe-design/styles/esm/parse-theme.js
import { TinyColor } from '@universe-design/colors';
export var TokenMetaType;
(function (TokenMetaType) {
    TokenMetaType["Raw"] = "Raw";
    TokenMetaType["MappingColor"] = "MappingColor";
    TokenMetaType["LinearGradient"] = "LinearGradient";
    TokenMetaType["ConicGradient"] = "ConicGradient";
    TokenMetaType["RadialGradient"] = "RadialGradient";
    TokenMetaType["Shadow"] = "Shadow";
    TokenMetaType["Length"] = "Length";
    TokenMetaType["FamilyName"] = "FamilyName";
})(TokenMetaType || (TokenMetaType = {}));
export function parseTheme(theme) {
    const plainTheme = {};
    const getRealValue = (meta, key) => {
        if (key) {
            const finalValue = plainTheme[key];
            if (finalValue !== undefined) {
                return finalValue;
            }
        }
        let ret;
        if (typeof meta === 'string') {
            ret = meta;
        }
        else if (meta.type === TokenMetaType.Raw) {
            ret = meta.raw;
        }
        else if (meta.type === TokenMetaType.MappingColor) {
            const color = new TinyColor(getRealValue(theme[meta.from], meta.from));
            if (meta.alpha !== undefined && meta.alpha !== 1) {
                ret = color.setAlpha(meta.alpha).toRgbString();
            }
            else {
                ret = color.toRgbString();
            }
        }
        else if (meta.type === TokenMetaType.LinearGradient) {
            ret = `linear-gradient(${meta.angle}, ${meta.stops
                .map(stop => `${getRealValue(stop.color)} ${stop.percentage * 100}%`)
                .join(', ')})`;
        }
        else if (meta.type === TokenMetaType.ConicGradient) {
            ret = `conic-gradient(${meta.angle}, ${meta.stops
                .map(stop => `${getRealValue(stop.color)} ${stop.percentage * 100}%`)
                .join(', ')})`;
        }
        else if (meta.type === TokenMetaType.RadialGradient) {
            ret = `radial-gradient(${meta.angle}, ${meta.stops
                .map(stop => `${getRealValue(stop.color)} ${stop.percentage * 100}%`)
                .join(', ')})`;
        }
        else if (meta.type === TokenMetaType.Length) {
            ret = meta.value;
        }
        else if (meta.type === TokenMetaType.FamilyName) {
            ret = meta.familyName.replace(/"/g, '\\"');
        }
        else {
            ret = meta.items
                .map(item => `${item.offset[0]}px ${item.offset[1]}px ${item.blur}px ${item.spread}px ${getRealValue(item.color)}`)
                .join(', ');
        }
        if (key) {
            plainTheme[key] = ret;
        }
        return ret;
    };
    Object.entries(theme).forEach(([key, meta]) => {
        getRealValue(meta, key);
    });
    return plainTheme;
}
