// @universe-design/styles/esm/preset-light.atom-raw.js
// autogen by ../scripts/gen-ts-from-json.ts
export const AI_cursor_raw = '71, 82, 230';
export const AI_icon_in_button_raw = '71, 82, 230';
export const B100_raw = '224, 233, 255';
export const B200_raw = '194, 212, 255';
export const B300_raw = '148, 180, 255';
export const B350_raw = '122, 162, 255';
export const B400_raw = '80, 131, 251';
export const B50_raw = '240, 244, 255';
export const B500_raw = '51, 109, 244';
export const B600_raw = '20, 86, 240';
export const B700_raw = '4, 66, 210';
export const B800_raw = '0, 47, 158';
export const B900_raw = '0, 34, 112';
export const C100_raw = '254, 226, 242';
export const C200_raw = '248, 196, 225';
export const C300_raw = '245, 152, 204';
export const C350_raw = '235, 120, 184';
export const C400_raw = '223, 88, 165';
export const C50_raw = '254, 240, 248';
export const C500_raw = '204, 57, 140';
export const C600_raw = '184, 40, 121';
export const C700_raw = '157, 21, 98';
export const C800_raw = '115, 7, 68';
export const C900_raw = '85, 12, 53';
export const G100_raw = '208, 245, 206';
export const G200_raw = '149, 229, 153';
export const G300_raw = '92, 209, 104';
export const G350_raw = '53, 189, 75';
export const G400_raw = '50, 166, 69';
export const G50_raw = '228, 250, 225';
export const G500_raw = '37, 136, 50';
export const G600_raw = '26, 117, 38';
export const G700_raw = '11, 96, 23';
export const G800_raw = '4, 67, 12';
export const G900_raw = '2, 44, 7';
export const I100_raw = '233, 234, 251';
export const I200_raw = '204, 207, 249';
export const I300_raw = '171, 176, 242';
export const I350_raw = '149, 155, 238';
export const I400_raw = '117, 125, 240';
export const I50_raw = '242, 243, 253';
export const I500_raw = '91, 101, 245';
export const I600_raw = '71, 82, 230';
export const I700_raw = '51, 61, 204';
export const I800_raw = '30, 39, 164';
export const I900_raw = '21, 27, 112';
export const L100_raw = '227, 240, 163';
export const L200_raw = '200, 221, 95';
export const L300_raw = '162, 193, 11';
export const L350_raw = '145, 173, 0';
export const L400_raw = '123, 146, 7';
export const L50_raw = '242, 248, 211';
export const L500_raw = '107, 127, 6';
export const L600_raw = '92, 109, 8';
export const L700_raw = '74, 88, 4';
export const L800_raw = '51, 61, 0';
export const L900_raw = '38, 46, 0';
export const N00_raw = '255, 255, 255';
export const N100_raw = '242, 243, 245';
export const N1000_raw = '0, 0, 0';
export const N200_raw = '239, 240, 241';
export const N300_raw = '222, 224, 227';
export const N350_raw = '208, 211, 214';
export const N400_raw = '187, 191, 196';
export const N50_raw = '245, 246, 247';
export const N500_raw = '143, 149, 158';
export const N600_raw = '100, 106, 115';
export const N650_raw = '81, 86, 93';
export const N700_raw = '55, 60, 67';
export const N800_raw = '43, 47, 54';
export const N900_raw = '31, 35, 41';
export const N950_raw = '15, 17, 20';
export const O100_raw = '254, 231, 205';
export const O200_raw = '254, 196, 139';
export const O300_raw = '255, 157, 76';
export const O350_raw = '255, 129, 26';
export const O400_raw = '237, 109, 12';
export const O50_raw = '255, 243, 229';
export const O500_raw = '194, 87, 5';
export const O600_raw = '164, 73, 4';
export const O700_raw = '133, 58, 5';
export const O800_raw = '100, 43, 2';
export const O900_raw = '59, 26, 2';
export const P100_raw = '239, 230, 254';
export const P200_raw = '220, 201, 253';
export const P300_raw = '200, 169, 252';
export const P350_raw = '183, 145, 250';
export const P400_raw = '159, 111, 241';
export const P50_raw = '245, 240, 254';
export const P500_raw = '141, 85, 237';
export const P600_raw = '122, 53, 240';
export const P700_raw = '97, 31, 214';
export const P800_raw = '72, 17, 166';
export const P900_raw = '47, 0, 128';
export const R100_raw = '254, 227, 226';
export const R200_raw = '253, 198, 196';
export const R300_raw = '248, 158, 155';
export const R350_raw = '255, 117, 112';
export const R400_raw = '245, 74, 69';
export const R50_raw = '254, 240, 240';
export const R500_raw = '226, 46, 40';
export const R600_raw = '192, 42, 38';
export const R700_raw = '161, 28, 23';
export const R800_raw = '116, 25, 21';
export const R900_raw = '89, 6, 3';
export const S100_raw = '255, 252, 163';
export const S200_raw = '255, 246, 122';
export const S300_raw = '255, 242, 88';
export const S350_raw = '255, 233, 40';
export const S400_raw = '229, 206, 0';
export const S50_raw = '255, 255, 219';
export const S500_raw = '194, 168, 0';
export const S600_raw = '143, 124, 0';
export const S700_raw = '92, 79, 0';
export const S800_raw = '66, 55, 0';
export const S900_raw = '44, 37, 2';
export const T100_raw = '196, 242, 236';
export const T200_raw = '111, 232, 216';
export const T300_raw = '51, 214, 192';
export const T350_raw = '45, 190, 171';
export const T400_raw = '16, 168, 147';
export const T50_raw = '226, 248, 245';
export const T500_raw = '15, 133, 117';
export const T600_raw = '6, 112, 98';
export const T700_raw = '4, 93, 81';
export const T800_raw = '3, 68, 59';
export const T900_raw = '2, 49, 42';
export const V100_raw = '249, 226, 249';
export const V200_raw = '243, 196, 243';
export const V300_raw = '229, 156, 229';
export const V350_raw = '222, 129, 222';
export const V400_raw = '207, 94, 207';
export const V50_raw = '252, 238, 252';
export const V500_raw = '191, 61, 191';
export const V600_raw = '166, 48, 166';
export const V700_raw = '135, 39, 135';
export const V800_raw = '106, 17, 106';
export const V900_raw = '82, 0, 82';
export const Vnext_bg_content_header_raw = '252, 252, 252';
export const Vnext_bg_sub_navigation_raw = '245, 245, 245';
export const Vnext_bg_thirdly_navigation_raw = '252, 252, 252';
export const W100_raw = '202, 239, 252';
export const W200_raw = '151, 220, 252';
export const W300_raw = '62, 195, 247';
export const W350_raw = '37, 176, 231';
export const W400_raw = '18, 149, 202';
export const W50_raw = '231, 248, 254';
export const W500_raw = '4, 127, 176';
export const W600_raw = '7, 106, 148';
export const W700_raw = '15, 88, 122';
export const W800_raw = '6, 65, 92';
export const W900_raw = '7, 43, 61';
export const Y100_raw = '250, 237, 194';
export const Y200_raw = '252, 223, 126';
export const Y300_raw = '250, 211, 85';
export const Y350_raw = '255, 198, 10';
export const Y400_raw = '217, 153, 4';
export const Y50_raw = '251, 244, 223';
export const Y500_raw = '173, 122, 3';
export const Y600_raw = '134, 91, 3';
export const Y700_raw = '111, 74, 1';
export const Y800_raw = '87, 54, 1';
export const Y900_raw = '56, 34, 1';
export const bg_base_raw = '242, 243, 245';
export const bg_body_overlay_raw = '245, 246, 247';
export const bg_body_raw = '255, 255, 255';
export const bg_content_base_raw = '248, 249, 250';
export const bg_filler_raw = '239, 240, 241';
export const bg_float_base_raw = '242, 243, 245';
export const bg_float_overlay_raw = '245, 246, 247';
export const bg_float_raw = '255, 255, 255';
export const bg_mask_raw = '0, 0, 0';
export const bg_pricolor_raw = '80, 131, 251';
export const bg_sub_navigation_raw = '245, 246, 247';
export const bg_tips_raw = '31, 35, 41';
export const colorful_blue_raw = '51, 109, 244';
export const colorful_carmine_raw = '223, 88, 165';
export const colorful_green_raw = '53, 189, 75';
export const colorful_indigo_raw = '91, 101, 245';
export const colorful_lime_raw = '145, 173, 0';
export const colorful_neutral_raw = '143, 149, 158';
export const colorful_orange_raw = '255, 129, 26';
export const colorful_purple_raw = '141, 85, 237';
export const colorful_red_raw = '245, 74, 69';
export const colorful_sunflower_raw = '255, 233, 40';
export const colorful_turquoise_raw = '45, 190, 171';
export const colorful_violet_raw = '191, 61, 191';
export const colorful_wathet_raw = '37, 176, 231';
export const colorful_yellow_raw = '255, 198, 10';
export const fill_disabled_raw = '187, 191, 196';
export const function_danger_100_raw = '254, 227, 226';
export const function_danger_200_raw = '253, 198, 196';
export const function_danger_300_raw = '248, 158, 155';
export const function_danger_350_raw = '255, 117, 112';
export const function_danger_400_raw = '245, 74, 69';
export const function_danger_50_raw = '254, 240, 240';
export const function_danger_500_raw = '226, 46, 40';
export const function_danger_600_raw = '192, 42, 38';
export const function_danger_700_raw = '161, 28, 23';
export const function_danger_800_raw = '116, 25, 21';
export const function_danger_900_raw = '89, 6, 3';
export const function_danger_content_default_raw = '245, 74, 69';
export const function_danger_content_hover_raw = '255, 117, 112';
export const function_danger_content_loading_raw = '248, 158, 155';
export const function_danger_content_pressed_raw = '226, 46, 40';
export const function_danger_fill_default_raw = '245, 74, 69';
export const function_danger_fill_hover_raw = '255, 117, 112';
export const function_danger_fill_loading_raw = '248, 158, 155';
export const function_danger_fill_pressed_raw = '226, 46, 40';
export const function_danger_fill_solid_01_raw = '254, 240, 240';
export const function_danger_fill_solid_02_raw = '254, 227, 226';
export const function_danger_fill_solid_03_raw = '253, 198, 196';
export const function_danger_on_danger_fill_raw = '255, 255, 255';
export const function_info_100_raw = '224, 233, 255';
export const function_info_200_raw = '194, 212, 255';
export const function_info_300_raw = '148, 180, 255';
export const function_info_350_raw = '122, 162, 255';
export const function_info_400_raw = '80, 131, 251';
export const function_info_50_raw = '240, 244, 255';
export const function_info_500_raw = '51, 109, 244';
export const function_info_600_raw = '20, 86, 240';
export const function_info_700_raw = '4, 66, 210';
export const function_info_800_raw = '0, 47, 158';
export const function_info_900_raw = '0, 34, 112';
export const function_info_content_default_raw = '20, 86, 240';
export const function_info_content_hover_raw = '51, 109, 244';
export const function_info_content_loading_raw = '148, 180, 255';
export const function_info_content_pressed_raw = '4, 66, 210';
export const function_info_fill_default_raw = '20, 86, 240';
export const function_info_fill_hover_raw = '51, 109, 244';
export const function_info_fill_loading_raw = '148, 180, 255';
export const function_info_fill_pressed_raw = '4, 66, 210';
export const function_info_fill_solid_01_raw = '240, 244, 255';
export const function_info_fill_solid_02_raw = '224, 233, 255';
export const function_info_fill_solid_03_raw = '194, 212, 255';
export const function_info_on_info_fill_raw = '255, 255, 255';
export const function_success_100_raw = '208, 245, 206';
export const function_success_200_raw = '149, 229, 153';
export const function_success_300_raw = '92, 209, 104';
export const function_success_350_raw = '53, 189, 75';
export const function_success_400_raw = '50, 166, 69';
export const function_success_50_raw = '228, 250, 225';
export const function_success_500_raw = '37, 136, 50';
export const function_success_600_raw = '26, 117, 38';
export const function_success_700_raw = '11, 96, 23';
export const function_success_800_raw = '4, 67, 12';
export const function_success_900_raw = '2, 44, 7';
export const function_success_content_default_raw = '50, 166, 69';
export const function_success_content_hover_raw = '53, 189, 75';
export const function_success_content_loading_raw = '92, 209, 104';
export const function_success_content_pressed_raw = '37, 136, 50';
export const function_success_fill_default_raw = '53, 189, 75';
export const function_success_fill_hover_raw = '92, 209, 104';
export const function_success_fill_loading_raw = '92, 209, 104';
export const function_success_fill_pressed_raw = '50, 166, 69';
export const function_success_fill_solid_01_raw = '228, 250, 225';
export const function_success_fill_solid_02_raw = '208, 245, 206';
export const function_success_fill_solid_03_raw = '149, 229, 153';
export const function_success_on_success_fill_raw = '255, 255, 255';
export const function_warning_100_raw = '254, 231, 205';
export const function_warning_200_raw = '254, 196, 139';
export const function_warning_300_raw = '255, 157, 76';
export const function_warning_350_raw = '255, 129, 26';
export const function_warning_400_raw = '237, 109, 12';
export const function_warning_50_raw = '255, 243, 229';
export const function_warning_500_raw = '194, 87, 5';
export const function_warning_600_raw = '164, 73, 4';
export const function_warning_700_raw = '133, 58, 5';
export const function_warning_800_raw = '100, 43, 2';
export const function_warning_900_raw = '59, 26, 2';
export const function_warning_content_default_raw = '237, 109, 12';
export const function_warning_content_hover_raw = '255, 129, 26';
export const function_warning_content_loading_raw = '255, 157, 76';
export const function_warning_content_pressed_raw = '194, 87, 5';
export const function_warning_fill_default_raw = '255, 129, 26';
export const function_warning_fill_hover_raw = '255, 157, 76';
export const function_warning_fill_loading_raw = '255, 157, 76';
export const function_warning_fill_pressed_raw = '237, 109, 12';
export const function_warning_fill_solid_01_raw = '255, 243, 229';
export const function_warning_fill_solid_02_raw = '254, 231, 205';
export const function_warning_fill_solid_03_raw = '254, 196, 139';
export const function_warning_on_warning_fill_raw = '255, 255, 255';
export const icon_disabled_raw = '187, 191, 196';
export const icon_n1_raw = '43, 47, 54';
export const icon_n2_raw = '100, 106, 115';
export const icon_n3_raw = '143, 149, 158';
export const illustration_blue_a_raw = '0, 34, 112';
export const illustration_blue_b_raw = '0, 34, 112';
export const illustration_blue_c_raw = '0, 34, 112';
export const illustration_blue_d_raw = '0, 34, 112';
export const illustration_blue_e_raw = '51, 109, 244';
export const illustration_neutral_a_raw = '255, 255, 255';
export const illustration_neutral_b_raw = '255, 255, 255';
export const illustration_neutral_d_raw = '143, 149, 158';
export const illustration_orange_raw = '255, 129, 26';
export const illustration_red_raw = '245, 74, 69';
export const illustration_static_black_raw = '0, 0, 0';
export const illustration_static_white_raw = '255, 255, 255';
export const illustration_turquoise_raw = '51, 214, 192';
export const illustration_wathet_raw = '62, 195, 247';
export const illustration_yellow_raw = '255, 198, 10';
export const inverse_primary_content_default_raw = '80, 131, 251';
export const inverse_primary_content_hover_raw = '122, 162, 255';
export const inverse_primary_content_pressed_raw = '51, 109, 244';
export const line_border_card_raw = '222, 224, 227';
export const line_border_component_raw = '208, 211, 214';
export const primary_content_default_raw = '20, 86, 240';
export const primary_content_hover_raw = '51, 109, 244';
export const primary_content_loading_raw = '148, 180, 255';
export const primary_content_pressed_raw = '4, 66, 210';
export const primary_fill_default_raw = '20, 86, 240';
export const primary_fill_hover_raw = '51, 109, 244';
export const primary_fill_loading_raw = '148, 180, 255';
export const primary_fill_pressed_raw = '4, 66, 210';
export const primary_fill_solid_01_raw = '240, 244, 255';
export const primary_fill_solid_02_raw = '224, 233, 255';
export const primary_fill_solid_03_raw = '194, 212, 255';
export const primary_on_primary_fill_raw = '255, 255, 255';
export const primary_pri_100_raw = '224, 233, 255';
export const primary_pri_200_raw = '194, 212, 255';
export const primary_pri_300_raw = '148, 180, 255';
export const primary_pri_350_raw = '122, 162, 255';
export const primary_pri_400_raw = '80, 131, 251';
export const primary_pri_50_raw = '240, 244, 255';
export const primary_pri_500_raw = '51, 109, 244';
export const primary_pri_600_raw = '20, 86, 240';
export const primary_pri_700_raw = '4, 66, 210';
export const primary_pri_800_raw = '0, 47, 158';
export const primary_pri_900_raw = '0, 34, 112';
export const shadow_default_raw = '31, 35, 41';
export const shadow_pri_raw = '20, 86, 240';
export const static_black_raw = '0, 0, 0';
export const static_white_hover_raw = '255, 255, 255';
export const static_white_pressed_raw = '255, 255, 255';
export const static_white_raw = '255, 255, 255';
export const text_caption_raw = '100, 106, 115';
export const text_disabled_raw = '187, 191, 196';
export const text_link_disabled_raw = '187, 191, 196';
export const text_link_hover_raw = '51, 109, 244';
export const text_link_loading_raw = '148, 180, 255';
export const text_link_normal_raw = '20, 86, 240';
export const text_link_pressed_raw = '4, 66, 210';
export const text_placeholder_raw = '143, 149, 158';
export const text_title_raw = '31, 35, 41';
