// @universe-design/react/lib/slider/handle.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SliderHandle = void 0;
var _clsx = _interopRequireDefault(require("clsx"));
var _react = _interopRequireWildcard(require("react"));
var _hooks = require("../config-provider/hooks");
var _tooltip = require("../tooltip");
var _util = require("../util");
var _useEventCallback = require("../hooks/useEventCallback");
const SliderHandle = props => {
  const {
    max,
    min,
    value,
    vertical,
    dragging,
    tooltipVisible: tooltipVisibleInProps,
    tooltipPlacement,
    tipFormatter,
    getTooltipPopupContainer
  } = props;
  const cssScope = (0, _hooks.useCssScope)('slider');
  const unSetVisible = typeof tooltipVisibleInProps === 'undefined';
  const [toolTipVisible, setToolTipVisible] = (0, _react.useState)();
  const udTooltipRef = (0, _react.useRef)(null);
  (0, _react.useEffect)(() => {
    if (unSetVisible && typeof dragging !== 'undefined') {
      setToolTipVisible(dragging);
    } else {
      setToolTipVisible(tooltipVisibleInProps);
    }
  }, [unSetVisible, tooltipVisibleInProps, dragging]);
  (0, _react.useEffect)(() => {
    var _udTooltipRef$current;
    // 1. 由于 `forcePopupAlign` 本身就是做了 throttle 的，所以我们用 `useEffect` 就可以了
    // 2. deps 数组先用 [value]，这种情况下如果 `min` 和 `max` 动态变化并不会触发 `forcePopupAlign`，但是大概率用户不会有拖动导致 `value` 变化的时候动态变化 `min` 和 `max` 的用法，先这样改，看看用户反馈
    (_udTooltipRef$current = udTooltipRef.current) === null || _udTooltipRef$current === void 0 ? void 0 : _udTooltipRef$current.forcePopupAlign();
  }, [value]);
  const handleTooltipVisibleChange = visible => {
    if (unSetVisible) {
      if (typeof dragging === 'undefined') {
        setToolTipVisible(visible);
      } else {
        setToolTipVisible(visible || dragging);
      }
    }
  };
  const handleHolderMouseDown = (0, _useEventCallback.useEventCallback)(ev => {
    // 解决 Safari 上拖动时光标从 pointer 变成 text 的问题
    ev.preventDefault();
  });
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, (() => {
    const v = value ? Math.round((value - min) / (max - min) * 100) : 0;
    const style = vertical ? {
      bottom: `${v}%`
    } : {
      left: `${v}%`
    };
    const handleCls = (0, _clsx.default)(cssScope('__handle'), {
      [cssScope('__handle--active')]: dragging
    });
    const title = (0, _util.isFunction)(tipFormatter) ? tipFormatter(value) : `${value}`;
    if ((unSetVisible || tooltipVisibleInProps) && tipFormatter !== null) {
      return /*#__PURE__*/_react.default.createElement(_tooltip.Tooltip, {
        title: title,
        visible: toolTipVisible,
        onVisibleChange: handleTooltipVisibleChange,
        getPopupContainer: getTooltipPopupContainer,
        placement: tooltipPlacement,
        udTriggerRef: udTooltipRef
      }, /*#__PURE__*/_react.default.createElement("div", {
        style: style,
        className: cssScope('__handle-holder'),
        onMouseDown: handleHolderMouseDown
      }, /*#__PURE__*/_react.default.createElement("div", {
        className: handleCls
      })));
    }
    return /*#__PURE__*/_react.default.createElement("div", {
      style: style,
      className: cssScope('__handle-holder'),
      onMouseDown: handleHolderMouseDown
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: handleCls
    }));
  })());
};
exports.SliderHandle = SliderHandle;
