// @universe-design/react/lib/slider/marks.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SliderMarks = void 0;
var _clsx = _interopRequireDefault(require("clsx"));
var _react = _interopRequireWildcard(require("react"));
var _hooks = require("../config-provider/hooks");
var _text = require("../text");
const SliderMarks = props => {
  const {
    max,
    min,
    marks,
    reverse,
    vertical,
    onMarkClick
  } = props;
  const cssScope = (0, _hooks.useCssScope)('slider');
  const markCls = (0, _clsx.default)(cssScope('__marks'), {
    [cssScope('__marks--vertical')]: vertical
  });
  const handleMarkClick = (e, value) => {
    e.stopPropagation();
    onMarkClick(value);
  };
  return marks ? /*#__PURE__*/_react.default.createElement("div", {
    className: markCls
  }, (() => {
    if (marks) {
      const keys = Object.keys(marks);
      return keys.map(parseFloat).sort((a, b) => a - b).map(key => {
        const item = marks[key];
        const isObject = typeof item === 'object' && ! /*#__PURE__*/(0, _react.isValidElement)(item);
        const label = isObject ? item.label : item;
        const style = isObject ? item.style : undefined;
        if (!label && label !== 0) {
          return null;
        }
        const cls = (0, _clsx.default)(cssScope('__marks-item'));
        const point = key;
        const mergedStyle = {
          ...style,
          ...(vertical ? {
            [reverse ? 'top' : 'bottom']: `${(point - min) / (max - min) * 100}%`
          } : {
            [reverse ? 'right' : 'left']: `${(point - min) / (max - min) * 100}%`
          })
        };
        return /*#__PURE__*/_react.default.createElement(_text.Text, {
          as: "div",
          variant: "body2",
          key: key,
          style: mergedStyle,
          className: cls,
          onMouseDown: e => handleMarkClick(e, point)
        }, label);
      });
    }
    return null;
  })()) : null;
};
exports.SliderMarks = SliderMarks;
