// @universe-design/styles/ts/ccm-ccm-doc/index.dark.atom-raw.ts
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
export const ccmtoken_doc_block_bg_area_raw = '166, 166, 166';
export const ccmtoken_doc_block_bg_hover_raw = '51, 112, 235';
export const ccmtoken_doc_block_bg_selected_raw = '51, 112, 235';
export const ccmtoken_doc_blockbackground_blue_raw = '25, 42, 76';
export const ccmtoken_doc_blockbackground_blue_solid_raw = '32, 62, 120';
export const ccmtoken_doc_blockbackground_gray_raw = '67, 67, 67';
export const ccmtoken_doc_blockbackground_gray_solid_raw = '95, 95, 95';
export const ccmtoken_doc_blockbackground_green_raw = '31, 71, 27';
export const ccmtoken_doc_blockbackground_green_solid_raw = '41, 107, 34';
export const ccmtoken_doc_blockbackground_netural_raw = '41, 41, 41';
export const ccmtoken_doc_blockbackground_netural_solid_raw = '55, 55, 55';
export const ccmtoken_doc_blockbackground_orange_raw = '87, 51, 10';
export const ccmtoken_doc_blockbackground_orange_solid_raw = '132, 81, 23';
export const ccmtoken_doc_blockbackground_purple_raw = '54, 29, 97';
export const ccmtoken_doc_blockbackground_purple_solid_raw = '77, 38, 145';
export const ccmtoken_doc_blockbackground_red_raw = '74, 29, 27';
export const ccmtoken_doc_blockbackground_red_solid_raw = '107, 40, 39';
export const ccmtoken_doc_blockbackground_yellow_raw = '92, 85, 20';
export const ccmtoken_doc_blockbackground_yellow_solid_raw = '135, 123, 16';
export const ccmtoken_doc_bottom_like_bg_hover_raw = '225, 180, 32';
export const ccmtoken_doc_bottom_like_bg_nomal_raw = '250, 200, 35';
export const ccmtoken_doc_button_bg_blue_raw = '25, 42, 76';
export const ccmtoken_doc_button_bg_blue_solid_raw = '76, 136, 255';
export const ccmtoken_doc_button_bg_carmine_raw = '92, 37, 66';
export const ccmtoken_doc_button_bg_carmine_solid_raw = '204, 71, 146';
export const ccmtoken_doc_button_bg_green_raw = '31, 71, 27';
export const ccmtoken_doc_button_bg_green_solid_raw = '59, 176, 46';
export const ccmtoken_doc_button_bg_neutral_raw = '55, 55, 55';
export const ccmtoken_doc_button_bg_neutral_solid_raw = '232, 232, 232';
export const ccmtoken_doc_button_bg_orange_raw = '87, 51, 10';
export const ccmtoken_doc_button_bg_orange_solid_raw = '222, 130, 24';
export const ccmtoken_doc_button_bg_purple_raw = '54, 29, 97';
export const ccmtoken_doc_button_bg_purple_solid_raw = '124, 74, 212';
export const ccmtoken_doc_button_bg_red_raw = '74, 29, 27';
export const ccmtoken_doc_button_bg_red_solid_raw = '204, 71, 67';
export const ccmtoken_doc_button_bg_wathet_raw = '23, 55, 66';
export const ccmtoken_doc_button_bg_wathet_solid_raw = '66, 189, 235';
export const ccmtoken_doc_calloutborder_blue_raw = '40, 81, 163';
export const ccmtoken_doc_calloutborder_green_raw = '51, 148, 40';
export const ccmtoken_doc_calloutborder_orange_raw = '189, 112, 23';
export const ccmtoken_doc_calloutborder_purple_raw = '101, 51, 189';
export const ccmtoken_doc_calloutborder_red_raw = '158, 56, 54';
export const ccmtoken_doc_calloutborder_yellow_raw = '199, 179, 0';
export const ccmtoken_doc_calloutborder_yellow_solid_raw = '232, 209, 0';
export const ccmtoken_doc_highlightcolor_bg_blue_soft_raw = '76, 136, 255';
export const ccmtoken_doc_highlightcolor_bg_blue_solid_raw = '76, 136, 255';
export const ccmtoken_doc_highlightcolor_bg_green_soft_raw = '84, 194, 72';
export const ccmtoken_doc_highlightcolor_bg_green_solid_raw = '84, 194, 72';
export const ccmtoken_doc_highlightcolor_bg_neutral_light_raw = '41, 41, 41';
export const ccmtoken_doc_highlightcolor_bg_orange_soft_raw = '242, 150, 44';
export const ccmtoken_doc_highlightcolor_bg_orange_solid_raw = '242, 150, 44';
export const ccmtoken_doc_highlightcolor_bg_purple_soft_raw = '151, 98, 245';
export const ccmtoken_doc_highlightcolor_bg_purple_solid_raw = '151, 98, 245';
export const ccmtoken_doc_highlightcolor_bg_red_soft_raw = '240, 91, 86';
export const ccmtoken_doc_highlightcolor_bg_red_solid_raw = '240, 91, 86';
export const ccmtoken_doc_highlightcolor_bg_sunflower_soft_raw = '245, 226, 49';
export const ccmtoken_doc_highlightcolor_bg_sunflower_solid_raw = '245, 226, 49';
export const ccmtoken_doc_highlightcolor_blue_soft_raw = '76, 136, 255';
export const ccmtoken_doc_highlightcolor_blue_solid_raw = '76, 136, 255';
export const ccmtoken_doc_highlightcolor_callout_neutral_light_raw = '41, 41, 41';
export const ccmtoken_doc_highlightcolor_green_soft_raw = '84, 194, 72';
export const ccmtoken_doc_highlightcolor_green_solid_raw = '84, 194, 72';
export const ccmtoken_doc_highlightcolor_neutral_light_raw = '67, 67, 67';
export const ccmtoken_doc_highlightcolor_neutral_soft_raw = '80, 80, 80';
export const ccmtoken_doc_highlightcolor_neutral_solid_raw = '95, 95, 95';
export const ccmtoken_doc_highlightcolor_orange_soft_raw = '242, 150, 44';
export const ccmtoken_doc_highlightcolor_orange_solid_raw = '242, 150, 44';
export const ccmtoken_doc_highlightcolor_purple_soft_raw = '151, 98, 245';
export const ccmtoken_doc_highlightcolor_purple_solid_raw = '151, 98, 245';
export const ccmtoken_doc_highlightcolor_red_soft_raw = '240, 91, 86';
export const ccmtoken_doc_highlightcolor_red_solid_raw = '240, 91, 86';
export const ccmtoken_doc_highlightcolor_sunflower_soft_raw = '245, 226, 49';
export const ccmtoken_doc_highlightcolor_sunflower_solid_raw = '245, 226, 49';
export const ccmtoken_doc_picture_toolbar_bg_raw = '41, 41, 41';
export const ccmtoken_doc_tableheader_bg_hover_raw = '80, 80, 80';
export const ccmtoken_doc_textcolor_bg_red_raw = '240, 91, 86';
export const ccmtoken_doc_textcolor_blue_raw = '112, 160, 255';
export const ccmtoken_doc_textcolor_callout_gray_raw = '235, 235, 235';
export const ccmtoken_doc_textcolor_gray_raw = '117, 117, 117';
export const ccmtoken_doc_textcolor_green_raw = '109, 209, 98';
export const ccmtoken_doc_textcolor_orange_raw = '245, 165, 74';
export const ccmtoken_doc_textcolor_purple_raw = '164, 114, 252';
export const ccmtoken_doc_textcolor_red_raw = '250, 120, 115';
export const ccmtoken_doc_textcolor_yellow_raw = '252, 212, 86';
export const ccmtoken_doc_toolbar_btn_bg_blue_hover_raw = '117, 164, 255';
export const ccmtoken_doc_toolbar_btn_bg_blue_pressed_raw = '117, 164, 255';
