// @universe-design/styles/ts/ccm-ccm-doc/index.dark.atom.ts
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
export const ccmtoken_doc_block_bg_area = 'rgba(166, 166, 166, 0.08)';
export const ccmtoken_doc_block_bg_hover = 'rgba(51, 112, 235, 0.16)';
export const ccmtoken_doc_block_bg_selected = 'rgba(51, 112, 235, 0.26)';
export const ccmtoken_doc_blockbackground_blue = '#192a4c';
export const ccmtoken_doc_blockbackground_blue_solid = '#203e78';
export const ccmtoken_doc_blockbackground_gray = '#434343';
export const ccmtoken_doc_blockbackground_gray_solid = '#5f5f5f';
export const ccmtoken_doc_blockbackground_green = '#1f471b';
export const ccmtoken_doc_blockbackground_green_solid = '#296b22';
export const ccmtoken_doc_blockbackground_netural = '#292929';
export const ccmtoken_doc_blockbackground_netural_solid = '#373737';
export const ccmtoken_doc_blockbackground_orange = '#57330a';
export const ccmtoken_doc_blockbackground_orange_solid = '#845117';
export const ccmtoken_doc_blockbackground_purple = '#361d61';
export const ccmtoken_doc_blockbackground_purple_solid = '#4d2691';
export const ccmtoken_doc_blockbackground_red = '#4a1d1b';
export const ccmtoken_doc_blockbackground_red_solid = '#6b2827';
export const ccmtoken_doc_blockbackground_yellow = '#5c5514';
export const ccmtoken_doc_blockbackground_yellow_solid = '#877b10';
export const ccmtoken_doc_bottom_like_bg_hover = '#e1b420';
export const ccmtoken_doc_bottom_like_bg_nomal = '#fac823';
export const ccmtoken_doc_button_bg_blue = '#192a4c';
export const ccmtoken_doc_button_bg_blue_solid = '#4c88ff';
export const ccmtoken_doc_button_bg_carmine = '#5c2542';
export const ccmtoken_doc_button_bg_carmine_solid = '#cc4792';
export const ccmtoken_doc_button_bg_green = '#1f471b';
export const ccmtoken_doc_button_bg_green_solid = '#3bb02e';
export const ccmtoken_doc_button_bg_neutral = '#373737';
export const ccmtoken_doc_button_bg_neutral_solid = '#e8e8e8';
export const ccmtoken_doc_button_bg_orange = '#57330a';
export const ccmtoken_doc_button_bg_orange_solid = '#de8218';
export const ccmtoken_doc_button_bg_purple = '#361d61';
export const ccmtoken_doc_button_bg_purple_solid = '#7c4ad4';
export const ccmtoken_doc_button_bg_red = '#4a1d1b';
export const ccmtoken_doc_button_bg_red_solid = '#cc4743';
export const ccmtoken_doc_button_bg_wathet = '#173742';
export const ccmtoken_doc_button_bg_wathet_solid = '#42bdeb';
export const ccmtoken_doc_calloutborder_blue = '#2851a3';
export const ccmtoken_doc_calloutborder_green = '#339428';
export const ccmtoken_doc_calloutborder_orange = '#bd7017';
export const ccmtoken_doc_calloutborder_purple = '#6533bd';
export const ccmtoken_doc_calloutborder_red = '#9e3836';
export const ccmtoken_doc_calloutborder_yellow = '#c7b300';
export const ccmtoken_doc_calloutborder_yellow_solid = '#e8d100';
export const ccmtoken_doc_highlightcolor_bg_blue_soft = 'rgba(76, 136, 255, 0.21)';
export const ccmtoken_doc_highlightcolor_bg_blue_solid = 'rgba(76, 136, 255, 0.35)';
export const ccmtoken_doc_highlightcolor_bg_green_soft = 'rgba(84, 194, 72, 0.17)';
export const ccmtoken_doc_highlightcolor_bg_green_solid = 'rgba(84, 194, 72, 0.28)';
export const ccmtoken_doc_highlightcolor_bg_neutral_light = 'rgba(41, 41, 41, 0.6)';
export const ccmtoken_doc_highlightcolor_bg_orange_soft = 'rgba(242, 150, 44, 0.17)';
export const ccmtoken_doc_highlightcolor_bg_orange_solid = 'rgba(242, 150, 44, 0.28)';
export const ccmtoken_doc_highlightcolor_bg_purple_soft = 'rgba(151, 98, 245, 0.24)';
export const ccmtoken_doc_highlightcolor_bg_purple_solid = 'rgba(151, 98, 245, 0.4)';
export const ccmtoken_doc_highlightcolor_bg_red_soft = 'rgba(240, 91, 86, 0.22)';
export const ccmtoken_doc_highlightcolor_bg_red_solid = 'rgba(240, 91, 86, 0.36)';
export const ccmtoken_doc_highlightcolor_bg_sunflower_soft = 'rgba(245, 226, 49, 0.13)';
export const ccmtoken_doc_highlightcolor_bg_sunflower_solid = 'rgba(245, 226, 49, 0.21)';
export const ccmtoken_doc_highlightcolor_blue_soft = 'rgba(76, 136, 255, 0.28)';
export const ccmtoken_doc_highlightcolor_blue_solid = 'rgba(76, 136, 255, 0.46)';
export const ccmtoken_doc_highlightcolor_callout_neutral_light = 'rgba(41, 41, 41, 0.6)';
export const ccmtoken_doc_highlightcolor_green_soft = 'rgba(84, 194, 72, 0.22)';
export const ccmtoken_doc_highlightcolor_green_solid = 'rgba(84, 194, 72, 0.36)';
export const ccmtoken_doc_highlightcolor_neutral_light = 'rgba(67, 67, 67, 0.4)';
export const ccmtoken_doc_highlightcolor_neutral_soft = 'rgba(80, 80, 80, 0.5)';
export const ccmtoken_doc_highlightcolor_neutral_solid = 'rgba(95, 95, 95, 0.6)';
export const ccmtoken_doc_highlightcolor_orange_soft = 'rgba(242, 150, 44, 0.22)';
export const ccmtoken_doc_highlightcolor_orange_solid = 'rgba(242, 150, 44, 0.36)';
export const ccmtoken_doc_highlightcolor_purple_soft = 'rgba(151, 98, 245, 0.31)';
export const ccmtoken_doc_highlightcolor_purple_solid = 'rgba(151, 98, 245, 0.51)';
export const ccmtoken_doc_highlightcolor_red_soft = 'rgba(240, 91, 86, 0.28)';
export const ccmtoken_doc_highlightcolor_red_solid = 'rgba(240, 91, 86, 0.47)';
export const ccmtoken_doc_highlightcolor_sunflower_soft = 'rgba(245, 226, 49, 0.16)';
export const ccmtoken_doc_highlightcolor_sunflower_solid = 'rgba(245, 226, 49, 0.26)';
export const ccmtoken_doc_picture_toolbar_bg = 'rgba(41, 41, 41, 0.7)';
export const ccmtoken_doc_tableheader_bg_hover = '#505050';
export const ccmtoken_doc_textcolor_bg_red = 'rgba(240, 91, 86, 0.35)';
export const ccmtoken_doc_textcolor_blue = '#70a0ff';
export const ccmtoken_doc_textcolor_callout_gray = '#ebebeb';
export const ccmtoken_doc_textcolor_gray = '#757575';
export const ccmtoken_doc_textcolor_green = '#6dd162';
export const ccmtoken_doc_textcolor_orange = '#f5a54a';
export const ccmtoken_doc_textcolor_purple = '#a472fc';
export const ccmtoken_doc_textcolor_red = '#fa7873';
export const ccmtoken_doc_textcolor_yellow = '#fcd456';
export const ccmtoken_doc_toolbar_btn_bg_blue_hover = 'rgba(117, 164, 255, 0.3)';
export const ccmtoken_doc_toolbar_btn_bg_blue_pressed = 'rgba(117, 164, 255, 0.4)';
