// @universe-design/styles/ts/ccm-ccm-doc/index.light.atom-raw.ts
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
export const ccmtoken_doc_block_bg_area_raw = '100, 106, 115';
export const ccmtoken_doc_block_bg_hover_raw = '20, 86, 240';
export const ccmtoken_doc_block_bg_selected_raw = '20, 86, 240';
export const ccmtoken_doc_blockbackground_blue_raw = '225, 234, 255';
export const ccmtoken_doc_blockbackground_blue_solid_raw = '186, 206, 253';
export const ccmtoken_doc_blockbackground_gray_raw = '222, 224, 227';
export const ccmtoken_doc_blockbackground_gray_solid_raw = '187, 191, 196';
export const ccmtoken_doc_blockbackground_green_raw = '217, 245, 214';
export const ccmtoken_doc_blockbackground_green_solid_raw = '183, 237, 177';
export const ccmtoken_doc_blockbackground_netural_raw = '242, 243, 245';
export const ccmtoken_doc_blockbackground_netural_solid_raw = '239, 240, 241';
export const ccmtoken_doc_blockbackground_orange_raw = '254, 234, 210';
export const ccmtoken_doc_blockbackground_orange_solid_raw = '254, 212, 164';
export const ccmtoken_doc_blockbackground_purple_raw = '236, 226, 254';
export const ccmtoken_doc_blockbackground_purple_solid_raw = '205, 178, 250';
export const ccmtoken_doc_blockbackground_red_raw = '253, 226, 226';
export const ccmtoken_doc_blockbackground_red_solid_raw = '251, 191, 188';
export const ccmtoken_doc_blockbackground_yellow_raw = '255, 255, 204';
export const ccmtoken_doc_blockbackground_yellow_solid_raw = '255, 252, 163';
export const ccmtoken_doc_bottom_like_bg_hover_raw = '255, 209, 59';
export const ccmtoken_doc_bottom_like_bg_nomal_raw = '255, 198, 10';
export const ccmtoken_doc_button_bg_blue_raw = '225, 234, 255';
export const ccmtoken_doc_button_bg_blue_solid_raw = '51, 112, 255';
export const ccmtoken_doc_button_bg_carmine_raw = '253, 221, 239';
export const ccmtoken_doc_button_bg_carmine_solid_raw = '241, 75, 169';
export const ccmtoken_doc_button_bg_green_raw = '217, 245, 214';
export const ccmtoken_doc_button_bg_green_solid_raw = '98, 210, 86';
export const ccmtoken_doc_button_bg_neutral_raw = '239, 240, 241';
export const ccmtoken_doc_button_bg_neutral_solid_raw = '43, 47, 54';
export const ccmtoken_doc_button_bg_orange_raw = '254, 234, 210';
export const ccmtoken_doc_button_bg_orange_solid_raw = '255, 165, 61';
export const ccmtoken_doc_button_bg_purple_raw = '236, 226, 254';
export const ccmtoken_doc_button_bg_purple_solid_raw = '147, 90, 246';
export const ccmtoken_doc_button_bg_red_raw = '253, 226, 226';
export const ccmtoken_doc_button_bg_red_solid_raw = '247, 105, 100';
export const ccmtoken_doc_button_bg_wathet_raw = '217, 243, 253';
export const ccmtoken_doc_button_bg_wathet_solid_raw = '20, 192, 255';
export const ccmtoken_doc_calloutborder_blue_raw = '130, 167, 252';
export const ccmtoken_doc_calloutborder_green_raw = '142, 224, 133';
export const ccmtoken_doc_calloutborder_orange_raw = '255, 186, 107';
export const ccmtoken_doc_calloutborder_purple_raw = '173, 130, 247';
export const ccmtoken_doc_calloutborder_red_raw = '249, 142, 139';
export const ccmtoken_doc_calloutborder_yellow_raw = '255, 246, 122';
export const ccmtoken_doc_calloutborder_yellow_solid_raw = '255, 242, 88';
export const ccmtoken_doc_highlightcolor_bg_blue_soft_raw = '240, 244, 255';
export const ccmtoken_doc_highlightcolor_bg_blue_solid_raw = '225, 234, 255';
export const ccmtoken_doc_highlightcolor_bg_green_soft_raw = '240, 251, 239';
export const ccmtoken_doc_highlightcolor_bg_green_solid_raw = '217, 245, 214';
export const ccmtoken_doc_highlightcolor_bg_neutral_light_raw = '245, 246, 247';
export const ccmtoken_doc_highlightcolor_bg_orange_soft_raw = '255, 245, 235';
export const ccmtoken_doc_highlightcolor_bg_orange_solid_raw = '254, 234, 210';
export const ccmtoken_doc_highlightcolor_bg_purple_soft_raw = '246, 241, 254';
export const ccmtoken_doc_highlightcolor_bg_purple_solid_raw = '236, 226, 254';
export const ccmtoken_doc_highlightcolor_bg_red_soft_raw = '254, 241, 241';
export const ccmtoken_doc_highlightcolor_bg_red_solid_raw = '253, 226, 226';
export const ccmtoken_doc_highlightcolor_bg_sunflower_soft_raw = '254, 255, 240';
export const ccmtoken_doc_highlightcolor_bg_sunflower_solid_raw = '255, 255, 204';
export const ccmtoken_doc_highlightcolor_blue_soft_raw = '186, 206, 253';
export const ccmtoken_doc_highlightcolor_blue_solid_raw = '78, 131, 253';
export const ccmtoken_doc_highlightcolor_callout_neutral_light_raw = '245, 246, 247';
export const ccmtoken_doc_highlightcolor_green_soft_raw = '183, 237, 177';
export const ccmtoken_doc_highlightcolor_green_solid_raw = '98, 210, 86';
export const ccmtoken_doc_highlightcolor_neutral_light_raw = '242, 243, 245';
export const ccmtoken_doc_highlightcolor_neutral_soft_raw = '222, 224, 227';
export const ccmtoken_doc_highlightcolor_neutral_solid_raw = '187, 191, 196';
export const ccmtoken_doc_highlightcolor_orange_soft_raw = '254, 212, 164';
export const ccmtoken_doc_highlightcolor_orange_solid_raw = '255, 165, 61';
export const ccmtoken_doc_highlightcolor_purple_soft_raw = '205, 178, 250';
export const ccmtoken_doc_highlightcolor_purple_solid_raw = '147, 90, 246';
export const ccmtoken_doc_highlightcolor_red_soft_raw = '251, 191, 188';
export const ccmtoken_doc_highlightcolor_red_solid_raw = '247, 105, 100';
export const ccmtoken_doc_highlightcolor_sunflower_soft_raw = '255, 246, 122';
export const ccmtoken_doc_highlightcolor_sunflower_solid_raw = '255, 233, 40';
export const ccmtoken_doc_picture_toolbar_bg_raw = '0, 0, 0';
export const ccmtoken_doc_tableheader_bg_hover_raw = '222, 224, 227';
export const ccmtoken_doc_textcolor_bg_red_raw = '245, 74, 69';
export const ccmtoken_doc_textcolor_blue_raw = '36, 91, 219';
export const ccmtoken_doc_textcolor_callout_gray_raw = '31, 35, 41';
export const ccmtoken_doc_textcolor_gray_raw = '143, 149, 158';
export const ccmtoken_doc_textcolor_green_raw = '46, 161, 33';
export const ccmtoken_doc_textcolor_orange_raw = '222, 120, 2';
export const ccmtoken_doc_textcolor_purple_raw = '100, 37, 208';
export const ccmtoken_doc_textcolor_red_raw = '216, 57, 49';
export const ccmtoken_doc_textcolor_yellow_raw = '220, 155, 4';
export const ccmtoken_doc_toolbar_btn_bg_blue_hover_raw = '20, 86, 240';
export const ccmtoken_doc_toolbar_btn_bg_blue_pressed_raw = '20, 86, 240';
