// @universe-design/styles/ts/ask/index.dark.atom.ts
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
export const asktoken_knowledge_btn_bg_copying_hover = '#4a2b10';
export const asktoken_knowledge_btn_bg_copying_normal = '#33210b';
export const asktoken_knowledge_btn_bg_downloading_hover = '#173b12';
export const asktoken_knowledge_btn_bg_downloading_normal = '#0e2b0a';
export const asktoken_knowledge_example_card_bg =
  'linear-gradient(to bottom, rgba(235, 235, 235, 0.05) 30%, rgba(235, 235, 235, 0) 100%)';
export const asktoken_knowledge_example_card_bg_2 = 'linear-gradient(to bottom, #212733 30%, rgba(33, 39, 51, 0) 100%)';
export const asktoken_knowledge_fill_file_blue = '#373737';
export const asktoken_knowledge_fill_file_green = '#373737';
export const asktoken_knowledge_fill_file_indigo = '#373737';
export const asktoken_knowledge_fill_file_neutral = '#373737';
export const asktoken_knowledge_fill_file_orange = '#373737';
export const asktoken_knowledge_fill_file_red = '#373737';
export const asktoken_knowledge_onboarding_text_gradient = 'linear-gradient(to bottom, #9499f7 0%, #a575fa 100%)';
export const asktoken_knowledge_rank_icon_bg = '#df9305';
export const asktoken_knowledge_rank_icon_outline = '#ffcb21';
export const asktoken_knowledge_scope_bg = '#373737';
