// @universe-design/react/src/layout/Layout.tsx
import type { ReactNode } from 'react';
import React, { forwardRef, useState } from 'react';
import clsx from 'clsx';
import { useCssScope } from '../config-provider/hooks';

import './style';

export interface BaseProps extends React.HTMLAttributes<HTMLDivElement> {
  hasSider?: boolean;
  children?: ReactNode;
}

export interface SiderControl {
  addSider: (id: string) => void;
  removeSider: (id: string) => void;
}

export interface LayoutContextProps {
  siderControl: SiderControl;
}
export const LayoutContext = React.createContext<LayoutContextProps>({
  siderControl: {
    addSider: () => null,
    removeSider: () => null,
  },
});

const Layout = forwardRef<HTMLElement, BaseProps>((props, ref) => {
  const { className, hasSider, children, ...restProps } = props;
  const [siders, setSiders] = useState<string[]>([]);
  const cssScope = useCssScope('layout');
  const ignoreHas = hasSider === false;
  const cls = clsx(
    {
      [cssScope()]: true,
      [cssScope('--has-sider')]: !ignoreHas && (hasSider || siders.length > 0),
    },
    className,
  );

  const getSiderControl = (): SiderControl => ({
    addSider: (id: string) => {
      setSiders(prev => [...prev, id]);
    },
    removeSider: (id: string) => {
      setSiders(prev => prev.filter(currentId => currentId !== id));
    },
  });

  return (
    <LayoutContext.Provider value={{ siderControl: getSiderControl() }}>
      <section ref={ref} className={cls} {...restProps}>
        {children}
      </section>
    </LayoutContext.Provider>
  );
});

export default Layout;
