// @universe-design/react/src/layout/Sider.tsx
import React, { forwardRef, useCallback, useContext, useEffect, useRef, useState } from 'react';
import clsx from 'clsx';
import { useControlledState } from '@universe-design/react-utilities';
import { useEffectOnce } from '../hooks';
import { LayoutContext } from './Layout';
import { useCssScope } from '../config-provider/hooks';
import { isFunction } from '../util';
import SideExpandOutlined from '@universe-design/icons-react/SideExpandOutlined';
import SideFoldOutlined from '@universe-design/icons-react/SideFoldOutlined';
import DisordeListOutlined from '@universe-design/icons-react/DisordeListOutlined';

import './style';

const dimensionMaxMap = {
  xs: '479.98px',
  sm: '575.98px',
  md: '767.98px',
  lg: '991.98px',
  xl: '1199.98px',
  xxl: '1599.98px',
};

export interface SiderContextProps {
  siderCollapsed: boolean;
  collapsedWidth: string | number;
}

export const SiderContext = React.createContext<SiderContextProps | null>(null);
if (__DEV__) {
  SiderContext.displayName = 'SiderContext';
}

export type CollapseType = 'clickTrigger' | 'responsive';

/**
 * @public
 */
export type SiderTheme = 'light' | 'dark';

/**
 * @public
 * Layout.Sider 的属性和方法
 */
export interface SiderProps extends React.HTMLAttributes<HTMLDivElement> {
  prefixCls?: string;
  /**
   * 是否可收起
   */
  collapsible?: boolean;
  /**
   * 当前收起状态
   */
  collapsed?: boolean;
  /**
   * 是否默认收起
   */
  defaultCollapsed?: boolean;
  /**
   * 翻转折叠提示箭头的方向，当  `Sider` 在右边时可以使用
   */
  reverseArrow?: boolean;
  /**
   * 展开-收起时的回调函数，有点击 trigger 以及响应式反馈两种方式可以触发
   */
  onCollapse?: (collapsed: boolean, type: CollapseType) => void;
  /**
   * 指定当 `collapsedWidth` 为 0 时出现的特殊 trigger 的样式
   */
  zeroWidthTriggerStyle?: React.CSSProperties;
  /**
   * 自定义 trigger，设置为 null 时隐藏 trigger
   */
  trigger?: React.ReactNode;
  /**
   * 宽度
   */
  width?: number | string;
  /**
   * 收缩宽度，设置为 0 会出现特殊 trigger
   */
  collapsedWidth?: number | string;
  /**
   * 触发响应式布局的断点
   */
  breakpoint?: 'xs' | 'sm' | 'md' | 'lg' | 'xl' | 'xxl';
  /**
   * 主题颜色
   */
  theme?: SiderTheme;
  /**
   * 触发响应式布局断点时的回调
   */
  onBreakpoint?: (broken: boolean) => void;
}

const generateId = (() => {
  let i = 0;
  return (prefix = '') => {
    i += 1;
    return `${prefix}${i}`;
  };
})();

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const isNumeric = (value: any): boolean => !isNaN(parseFloat(value)) && isFinite(value);

// eslint-disable-next-line max-lines-per-function
const Sider = forwardRef<HTMLElement, SiderProps>((props, ref) => {
  const {
    theme,
    style,
    className,
    width = 200,
    trigger,
    collapsed,
    collapsedWidth = 64,
    defaultCollapsed,
    collapsible,
    reverseArrow,
    zeroWidthTriggerStyle,
    breakpoint,
    children,
    onCollapse,
    onBreakpoint,
    ...othereProps
  } = props;
  const cssScope = useCssScope('layout__sider');
  const prefixCls = cssScope();
  const mqlRef = useRef<MediaQueryList>();
  const { siderControl } = useContext(LayoutContext);
  const [below, setBelow] = useState<boolean>(false);
  const [ignoreState, setIgnoreState] = useState<boolean>(false);
  const [collapsedState, setCollapsedState] = useControlledState<boolean>(Boolean(defaultCollapsed), collapsed);

  const setCollapsed = useCallback(
    (collapsedValue: boolean, type: CollapseType) => {
      setCollapsedState(collapsedValue);
      isFunction(onCollapse) && onCollapse(collapsedValue, type);
    },
    [onCollapse, setCollapsedState],
  );

  const responsiveHandler = useCallback(
    (mql: MediaQueryListEvent | MediaQueryList) => {
      setBelow(mql.matches);
      isFunction(onBreakpoint) && onBreakpoint(mql.matches);
      if (!mql.matches) {
        setIgnoreState(false);
      }
      if (!ignoreState && collapsedState !== mql.matches) {
        setCollapsed(mql.matches, 'responsive');
      }
    },
    [collapsedState, ignoreState, onBreakpoint, setCollapsed, setIgnoreState],
  );

  const toggle = () => {
    setCollapsed(!collapsedState, 'clickTrigger');
    setIgnoreState(true);
  };

  if (!mqlRef.current && typeof window !== 'undefined' && window.matchMedia && breakpoint) {
    mqlRef.current = window.matchMedia(`(max-width: ${dimensionMaxMap[breakpoint]})`);
    (mqlRef.current.addEventListener || mqlRef.current.addListener)('change', responsiveHandler);
    responsiveHandler(mqlRef.current);
  }

  useEffectOnce(() => {
    const uuid = generateId(cssScope());
    siderControl.addSider(uuid);
    return () => {
      siderControl.removeSider(uuid);
    };
  });

  useEffect(() => {
    let mql: MediaQueryList | null = null;
    if (!mql && typeof window !== 'undefined' && window.matchMedia && breakpoint) {
      mql = window.matchMedia(`(max-width: ${dimensionMaxMap[breakpoint]})`);
      if (mql.addEventListener) {
        mql.addEventListener('change', responsiveHandler);
      } else {
        mql.addListener(responsiveHandler);
      }
      responsiveHandler(mql);
    }
    return () => {
      if (mql) {
        if (mql.removeEventListener) {
          mql.removeEventListener('change', responsiveHandler);
        } else {
          mql.removeListener(responsiveHandler);
        }
        mql = null;
      }
    };
  }, [breakpoint, responsiveHandler]);

  const rawWidth = collapsedState ? collapsedWidth : width;
  const siderWidth = isNumeric(rawWidth) ? `${rawWidth}px` : String(rawWidth);
  const zeroWidthTrigger =
    parseFloat(String(collapsedWidth || 0)) === 0 ? (
      <span
        onClick={toggle}
        className={clsx(
          cssScope('--zero-width-trigger'),
          cssScope(`--zero-width-trigger-${reverseArrow ? 'right' : 'left'}`),
        )}
        style={zeroWidthTriggerStyle}
      >
        {trigger || <DisordeListOutlined />}
      </span>
    ) : null;

  const SiderStyle = {
    ...style,
    flex: `0 0 ${siderWidth}`,
    maxWidth: siderWidth,
    minWidth: siderWidth,
    width: siderWidth,
    '--collapsedWidth': typeof collapsedWidth === 'number' ? `${collapsedWidth}px` : collapsedWidth,
  };

  const siderCls = clsx(className, prefixCls, {
    [cssScope(`__${theme}`)]: theme,
    [cssScope('--collapsed')]: Boolean(collapsedState),
    [cssScope('--has-trigger')]: collapsible && trigger !== null && !zeroWidthTrigger,
    [cssScope('--below')]: Boolean(below),
    [cssScope('--zero-width')]: parseFloat(siderWidth) === 0,
  });

  const iconObj = {
    expanded: reverseArrow ? <SideFoldOutlined /> : <SideExpandOutlined />,
    collapsed: reverseArrow ? <SideExpandOutlined /> : <SideFoldOutlined />,
  };
  const status = collapsedState ? 'collapsed' : 'expanded';
  const defaultTrigger = iconObj[status];

  const renderTrigger = () => {
    if (collapsible || (below && zeroWidthTrigger && trigger !== null)) {
      return (
        zeroWidthTrigger || (
          <div className={`${prefixCls}-trigger`} onClick={toggle} style={{ width: siderWidth }}>
            {trigger || defaultTrigger}
          </div>
        )
      );
    }
    return null;
  };

  return (
    <SiderContext.Provider
      value={{
        siderCollapsed: collapsedState,
        collapsedWidth,
      }}
    >
      <aside ref={ref} style={SiderStyle} className={siderCls} {...othereProps}>
        <div className={cssScope('__children')}>{children}</div>
        {renderTrigger()}
      </aside>
    </SiderContext.Provider>
  );
});

export default Sider;
