// @universe-design/react-utilities/lib-commonjs/dom/resizeObservable.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.resizeObservable = exports.ResizeObservable = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _resizeObserver = require("@juggle/resize-observer");
const isSafari = () => typeof navigator !== 'undefined' && /^((?!chrome|android).)*safari/i.test(navigator.userAgent);
/**
 * A wrapped ResizeObservable class. Every element has their own resize observable, using WeakMap inside.
 *
 * @internal
 */
class ResizeObservable {
  constructor() {
    (0, _defineProperty2.default)(this, "_roWM", new WeakMap());
  }
  get ResizeObserver() {
    return typeof ResizeObserver !== 'undefined' && !isSafari() ? ResizeObserver : _resizeObserver.ResizeObserver;
  }
  subscribe(node, callback) {
    var _callbackResizeObserv;
    let callbackResizeObserverWeakMap = this._roWM.get(node);
    if ((_callbackResizeObserv = callbackResizeObserverWeakMap) !== null && _callbackResizeObserv !== void 0 && _callbackResizeObserv.get(callback)) {
      // 不能重复绑定同一个函数多次
      return () => void 0;
    }
    const ro = new this.ResizeObserver(entries => {
      for (const entry of entries) {
        if (entry.target === node) {
          callback(entry);
          break;
        }
      }
    });
    ro.observe(node, {
      box: 'border-box'
    }); // 这里先统一用 border-box，之后如果有需求可以开放其他的 box 类型，但是需要注意 WeakMap 逻辑也要改动
    if (!callbackResizeObserverWeakMap) {
      callbackResizeObserverWeakMap = new WeakMap();
      this._roWM.set(node, callbackResizeObserverWeakMap);
    }
    callbackResizeObserverWeakMap.set(callback, ro);
    return () => {
      this.unsubscribe(node, callback);
    };
  }
  unsubscribe(node, callback) {
    const callbackResizeObserverWeakMap = this._roWM.get(node);
    if (!callbackResizeObserverWeakMap) {
      return;
    }
    const ro = callbackResizeObserverWeakMap.get(callback);
    if (!ro) {
      return;
    }
    ro.disconnect();
    callbackResizeObserverWeakMap.delete(callback);
  }
}

/**
 * Module scope singleton of wrapped ResizeObservable.
 *
 * @internal
 */
exports.ResizeObservable = ResizeObservable;
const resizeObservable = exports.resizeObservable = new ResizeObservable();
