// @byted-service/netpool/lib/tcp/operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TCPOperation = void 0;
const net_1 = require("net");
const util_1 = require("util");
const helper_1 = require("../helper");
const consts_1 = require("../consts");
const idCreator = (0, helper_1.genIdCreator)('tcpsocket');
/**
 * Global clock
 * 全局时钟
 */
let now = Date.now();
setInterval(() => (now = Date.now()), 1000).unref();
class TCPOperation {
    constructor(options) {
        this.id = '-';
        this.waitingDestroySocketNum = 0;
        this.connectOptions = options.connectOptions;
        this.connectTimeout =
            typeof options.connectTimeout === 'number' ||
                options.connectTimeout === null
                ? options.connectTimeout
                : consts_1.defaultTCPPoolOptions.connectTimeout;
        this.socketDataCheckInterval =
            options.socketDataCheckInterval ||
                consts_1.defaultTCPPoolOptions.socketDataCheckInterval;
        this.maxSocketDataCheckTime =
            options.maxSocketDataCheckTime ||
                consts_1.defaultTCPPoolOptions.maxSocketDataCheckTime;
        this.maxWaitingDestroySocketNum =
            options.maxWaitingDestroySocketNum ||
                consts_1.defaultTCPPoolOptions.maxWaitingDestroySocketNum;
        this.maxBufferSize =
            options.maxBufferSize || consts_1.defaultTCPPoolOptions.maxBufferSize;
        this.maxSocketSurvivalTime =
            options.maxSocketSurvivalTime ||
                consts_1.defaultTCPPoolOptions.maxSocketSurvivalTime;
        this.maxSocketAge =
            options.maxSocketAge || consts_1.defaultTCPPoolOptions.maxSocketAge;
        this.customCreate = options.create;
        this.customDestroy = options.destroy;
        this.customValidate = options.validate;
        this.customReset = options.reset;
        this.logger = options.logger || consts_1.defaultLogger;
    }
    create() {
        return new Promise((resolve, reject) => {
            const socket = (0, net_1.connect)(this.connectOptions);
            socket.id = idCreator();
            socket.age = 0;
            socket.createdAt = now;
            socket.on('error', consts_1.noop);
            socket.on('connect', onConnect);
            socket.on('error', onError);
            let timeoutId = null;
            if (this.connectTimeout) {
                timeoutId = setTimeout(onTimeout, this.connectTimeout);
            }
            const self = this;
            function cleanup() {
                socket.removeListener('connect', onConnect);
                socket.removeListener('error', onError);
                if (timeoutId) {
                    clearTimeout(timeoutId);
                }
            }
            function onConnect() {
                socket.setKeepAlive(true);
                if (self.customCreate) {
                    self.customCreate(socket);
                }
                cleanup();
                resolve(socket);
            }
            function onError(err) {
                try {
                    socket.destroy();
                }
                catch (err) {
                    // do nothing
                }
                cleanup();
                reject(err);
            }
            function onTimeout() {
                try {
                    socket.destroy();
                }
                catch (err) {
                    // do nothing
                }
                const err = new Error((0, util_1.format)('connect ETIMEDOUT. Timeout: %dms. Options: %O', self.connectTimeout, self.connectOptions));
                // @ts-ignore
                err.errorno = 'ETIMEDOUT';
                // @ts-ignore
                err.code = 'ETIMEDOUT';
                // @ts-ignore
                err.syscall = 'connect';
                cleanup();
                reject(err);
            }
        });
    }
    async destroy(socket) {
        if (!socket.destroyed) {
            this.waitingDestroySocketNum++;
            const start = now;
            while (!socket.destroyed &&
                (socket.bufferSize || socket.writableLength) > 0 &&
                this.waitingDestroySocketNum <= this.maxWaitingDestroySocketNum) {
                if (now - start >= this.maxSocketDataCheckTime) {
                    this.logger.warn('[netpool] [%s/%s/%s] Reach options.maxSocketDataCheckTime: %dms. Try end and destroy socket %s, which may cause data loss', this.id, socket.id, '-', this.maxSocketDataCheckTime, socket.id);
                    try {
                        socket.end();
                    }
                    catch (err) {
                        // do nothing
                    }
                    break;
                }
                await new Promise((resolve) => {
                    setTimeout(resolve, this.socketDataCheckInterval).unref();
                });
            }
            if (this.waitingDestroySocketNum > this.maxWaitingDestroySocketNum) {
                this.logger.warn('[netpool] [%s/%s/%s] Reach options.maxWaitingDestroySocketNum: %d. Try destroy socket %s, which may cause data loss. ' +
                    'You can increase options.maxWaitingDestroySocketNum to avoid it. ' +
                    'But be careful not to reach the maximum number of open file handles', this.id, socket.id, '-', this.maxWaitingDestroySocketNum, socket.id);
            }
            socket.destroy();
            this.waitingDestroySocketNum--;
        }
        socket.unref();
        if (this.customDestroy) {
            this.customDestroy(socket);
        }
    }
    validate(socket) {
        /**
         * Check whether the socket has a reason for destroy
         * 检查 socket 是否有销毁原因
         */
        if (socket.destroyReason) {
            return false;
        }
        /**
         * Check whether the socket is destroyed
         * 检查 socket 是否已销毁
         */
        if (socket.destroyed) {
            socket.destroyReason = 'socket already destroyed';
            return false;
        }
        /**
         * Check whether the socket is readable && writable
         * 检查 socket 是否双工
         *
         * https://github.com/luin/ioredis/blob/master/lib/redis/index.ts#L713
         */
        if (socket.readable === false || socket.writable === false) {
            socket.destroyReason = "socket isn't readable or writable";
            return false;
        }
        else if (socket.readable === undefined) {
            const readableState = socket._readableState;
            /* istanbul ignore next */
            if (readableState && readableState.ended) {
                socket.destroyReason = "socket isn't readable";
                return false;
            }
            const writableState = socket._writableState;
            /* istanbul ignore next */
            if (writableState && writableState.ended) {
                socket.destroyReason = "socket isn't writable";
                return false;
            }
        }
        // <!-- Socket recycling strategy -->
        /**
         * Check socket write buffer capacity
         * 检查 socket 写 buffer 容量
         */
        if ((socket.bufferSize || socket.writableLength) >= this.maxBufferSize) {
            /**
             * This situation may be caused by writing too much data at one time, or the system writing is too slow
             * 出现这种情况可能是因为一次性写入过多数据，或者是系统写入过慢
             *
             * So, in order to reduce the risk of data loss, the socket is directly discarded, and in the destroy phase, a special timing task is used to track
             * 为了降低数据丢失风险，直接抛弃该 socket，在 destroy 阶段，由专门的定时任务去跟踪
             */
            socket.destroyReason = `socket bufferSize is greate than ${this.maxBufferSize} bytes`;
            return false;
        }
        /**
         * Check socket survival time
         * 检查 socket 存活时间
         */
        if (socket.createdAt &&
            now - socket.createdAt >= this.maxSocketSurvivalTime) {
            socket.destroyReason = `socket survival time is greate than ${this.maxSocketSurvivalTime}ms`;
            return false;
        }
        /**
         * Check socket available times
         * 检查 socket 使用年龄
         */
        if (socket.age && socket.age >= this.maxSocketAge) {
            socket.destroyReason = `socket use age is greate than ${this.maxSocketAge}`;
            return false;
        }
        // <!-- Socket recycling strategy -->
        /**
         * Custom verification rules
         * 自定义校验规则
         */
        if (this.customValidate) {
            const result = this.customValidate(socket);
            if (result !== true) {
                socket.destroyReason =
                    socket.destroyReason ||
                        "socket didn't pass custom validate";
            }
            return result;
        }
        return true;
    }
    reset(socket, stage) {
        if (socket.destroyed) {
            return socket;
        }
        if (stage === 'acquire') {
            socket.age++;
        }
        if (this.customReset) {
            return this.customReset(socket);
        }
        return socket;
    }
}
exports.TCPOperation = TCPOperation;
