// @universe-design/react/es/radio/group.js
import React, { forwardRef } from 'react';
import clsx from 'clsx';
import { useControlledState } from '@universe-design/react-utilities';
import Radio from './radio';
import { useComponentSize, useCssScope } from '../config-provider/hooks';
import { RadioGroupContextProvider } from './context';
import { isFunction, getDefaultSizeAbbr } from '../util';
import './style';
/**
 * @public
 */

/**
 * @public
 */

/**
 * 单选框组合，用于包裹一组单选器
 */

/**
 * 如果有 radio button child，父级 optionType 就要是 button
 * https://stackoverflow.com/questions/21252551/apply-style-to-parent-if-it-has-child-with-css
 */
function useRadioGroupButtonOptionType(_ref) {
  let {
    optionType: optionTypeProp = 'default'
  } = _ref;
  const [optionTypeRegistry, setButtonRegistry] = React.useState(new Map());
  const registerBOT = React.useCallback(value => {
    setButtonRegistry(s => new Map([...s.entries(), [value, 'button']]));
  }, []);
  const unregisterBOT = React.useCallback(value => {
    setButtonRegistry(s => {
      s.delete(value);
      return new Map([...s.entries()]);
    });
  }, []);
  const optionType = optionTypeRegistry.size ? 'button' : optionTypeProp;
  return {
    registerBOT,
    unregisterBOT,
    optionType
  };
}
const RadioGroup = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    className,
    options,
    direction = 'horizontal',
    // buttonStyle = 'outline',
    disabled,
    children,
    size,
    style,
    id,
    onMouseEnter,
    onMouseLeave,
    prefixCls
  } = props;
  const ctxComponentSize = useComponentSize();
  const mergedSize = size !== null && size !== void 0 ? size : ctxComponentSize;
  const groupCssScope = useCssScope('radio-group', prefixCls);
  const radioButtonScope = useCssScope('radio-button', prefixCls)();
  const [value, setValue] = useControlledState(props.defaultValue, props.value, {
    isControlledOverride: 'value' in props
  });
  const {
    registerBOT,
    unregisterBOT,
    optionType
  } = useRadioGroupButtonOptionType(props);
  return /*#__PURE__*/React.createElement(RadioGroupContextProvider, {
    value: {
      onChange: ev => {
        const val = ev.target.value;
        // 在 group 模式下，如果没有提供 val，不会触发 onChange。
        if (val === undefined) {
          return;
        }
        setValue(val);
        const {
          onChange
        } = props;
        isFunction(onChange) && val !== value && onChange(val, ev);
      },
      value,
      disabled: props.disabled,
      name: props.name,
      error: props.error,
      registerBOT,
      unregisterBOT,
      size: mergedSize
    }
  }, (() => {
    let renderRadio = children;
    if (options && options.length > 0) {
      const radioPrefixCls = optionType === 'button' ? radioButtonScope : undefined;
      renderRadio = options.map(option => {
        if (typeof option === 'string') {
          return /*#__PURE__*/React.createElement(Radio, {
            key: option,
            disabled: disabled,
            value: option,
            checked: value === option,
            prefixCls: radioPrefixCls,
            optionType: optionType
          }, option);
        }
        return /*#__PURE__*/React.createElement(Radio, {
          key: `radio-option-${option.value}`,
          disabled: option.disabled || disabled,
          value: option.value,
          checked: value === option.value,
          style: option.style,
          className: option.className,
          prefixCls: radioPrefixCls,
          optionType: optionType
        }, option.label);
      });
    }
    const componentSize = getDefaultSizeAbbr(mergedSize);
    const classNames = clsx(groupCssScope(), optionType === 'button' && groupCssScope('-button'), groupCssScope('--mode-solid'),
    // 目前我们的设计只有 solid mode 一种样式
    groupCssScope(`--direction-${direction}`), {
      [groupCssScope(`--size-${componentSize}`)]: componentSize
    }, className);
    return /*#__PURE__*/React.createElement("div", {
      className: classNames,
      style: style,
      onMouseEnter: onMouseEnter,
      onMouseLeave: onMouseLeave,
      id: id,
      ref: ref
    }, renderRadio);
  })());
});
export default /*#__PURE__*/React.memo(RadioGroup);
