// @universe-design/react/es/radio/radio.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.uf9s3yw{margin-left:8px;}
import u5msrnt from "./../.stylex/u5msrnt";
//.uokg9q4{overflow-wrap:break-word;}
import u1pdmi33 from "./../.stylex/u1pdmi33";
//.u132xexn{max-width:240px;}
import u1tqlmkb from "./../.stylex/u1tqlmkb";
//.u1sbtcvk{padding-top:5px;}
import u15pdtm from "./../.stylex/u15pdtm";
//.u11qrl6u{padding-right:12px;}
import u143d86t from "./../.stylex/u143d86t";
//.udghr9{padding-bottom:5px;}
import u1y2b6fc from "./../.stylex/u1y2b6fc";
//.ujlbh76{padding-left:12px;}
import urjm3ap from "./../.stylex/urjm3ap";
//.u1sl3k7w{padding-top:1px;}
import u1rs2opo from "./../.stylex/u1rs2opo";
//.u1brlhvm{padding-bottom:1px;}
import uh8azbh from "./../.stylex/uh8azbh";
//.uq1loh5{padding-top:2px;}
import upgelb9 from "./../.stylex/upgelb9";
//.u19lj068{padding-right:8px;}
import u441rjq from "./../.stylex/u441rjq";
//.uutqtb8{padding-bottom:2px;}
import u1vsjzop from "./../.stylex/u1vsjzop";
//.u177v4lu{padding-left:8px;}
import um3isuf from "./../.stylex/um3isuf";
import React, { useRef, forwardRef, useImperativeHandle } from 'react';
import clsx from 'clsx';
import stylex from '@universe-design/stylex';
import { useControlledState } from '@universe-design/react-utilities';
import { cssFontMedium, cssTypographyBody2, cssTypographyBody0, cssBlockBase } from '../stylex';
import useRadioGroupContext from './useRadioGroupContext';
import { useCssScope } from '../config-provider/hooks';
import { isFunction, omit } from '../util';
import { Text } from '../text';
import './style';
const RADIO_LABEL_MARGIN = 8;
const RADIO_BUTTON_MAX_WIDTH = 240;
export const Radio = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    prefixCls,
    className,
    role = 'radio',
    checked,
    style,
    onChange,
    onMouseEnter,
    onMouseLeave,
    children,
    value,
    disabled,
    name,
    defaultChecked,
    error,
    optionType = 'default',
    ...restProps
  } = props;
  const context = useRadioGroupContext();
  const inputRef = useRef(null);
  const cssScope = useCssScope('radio', prefixCls);
  const rootPrefixCls = cssScope();
  const isButton = optionType === 'button';
  const [checkedState, setCheckedState] = useControlledState(Boolean(defaultChecked), context ? context.value === value : Boolean(checked), {
    isControlledOverride: Boolean(context) || 'checked' in props
  });
  useImperativeHandle(ref, () => inputRef.current);
  let disableState = disabled;
  let nameState = name;
  let errorState = error;
  if (context) {
    disableState = disabled || context.disabled;
    nameState = context.name;
    errorState = error || context.error;
  }
  const wrapperCls = (() => {
    let cssWrapperSize;
    if ((context === null || context === void 0 ? void 0 : context.size) === 'small') {
      cssWrapperSize = {
        "props": {
          "z8tnut": upgelb9,
          "z189sj": u441rjq,
          "Byoj8tv": u1vsjzop,
          "uwmqm3": um3isuf
        }
      };
    } else if ((context === null || context === void 0 ? void 0 : context.size) === 'medium') {
      cssWrapperSize = {
        "props": {
          "z8tnut": u1rs2opo,
          "z189sj": u441rjq,
          "Byoj8tv": uh8azbh,
          "uwmqm3": um3isuf
        }
      };
    } else if ((context === null || context === void 0 ? void 0 : context.size) === 'large') {
      cssWrapperSize = {
        "props": {
          "z8tnut": u15pdtm,
          "z189sj": u143d86t,
          "Byoj8tv": u1y2b6fc,
          "uwmqm3": urjm3ap
        }
      };
    }
    return clsx(`${rootPrefixCls}__wrapper`, {
      [`${rootPrefixCls}__wrapper--checked`]: checkedState,
      [`${rootPrefixCls}__wrapper--disable`]: disableState,
      [`${rootPrefixCls}--error`]: errorState
    }, className, stylex(cssTypographyBody0,
    // 使用 button style 的时候选中需要加粗
    isButton && checkedState && cssFontMedium, isButton && cssWrapperSize));
  })();
  const sxLabelContent = (() => {
    let cssContentSize;
    if ((context === null || context === void 0 ? void 0 : context.size) === 'small') {
      cssContentSize = cssTypographyBody2;
    } else if ((context === null || context === void 0 ? void 0 : context.size) === 'medium') {
      cssContentSize = cssTypographyBody0;
    } else if ((context === null || context === void 0 ? void 0 : context.size) === 'large') {
      cssContentSize = cssTypographyBody0;
    }
    const sxArray = [cssBlockBase];
    if (isButton) {
      sxArray.push(cssContentSize, {
        "props": {
          "B2u0y6b": u1tqlmkb
        }
      });
    } else {
      sxArray.push({
        "props": {
          "Frg6f3": u5msrnt,
          "Btd35i7": u1pdmi33
        }
      });
    }
    return sxArray;
  })();
  return /*#__PURE__*/React.createElement("label", {
    className: wrapperCls,
    style: style,
    onMouseEnter: onMouseEnter,
    onMouseLeave: onMouseLeave
  }, /*#__PURE__*/React.createElement("span", {
    className: rootPrefixCls
  }, /*#__PURE__*/React.createElement("input", _extends({
    ref: inputRef,
    className: clsx(cssScope('__input')),
    type: "radio",
    role: role,
    checked: checkedState,
    onChange: function (e) {
      const eventParams = {
        target: {
          type: 'radio',
          checked: e.target.checked,
          value
        },
        stopPropagation() {
          e.stopPropagation();
        },
        preventDefault() {
          e.preventDefault();
        },
        nativeEvent: e.nativeEvent,
        reactEvent: e
      };
      if (context && isFunction(context.onChange)) {
        context.onChange(eventParams);
      } else {
        setCheckedState(e.target.checked);
        isFunction(onChange) && onChange(eventParams);
      }
    },
    "aria-checked": checkedState,
    disabled: disableState,
    name: nameState
  }, omit(restProps, ['error']))), /*#__PURE__*/React.createElement("span", {
    className: cssScope('__wallpaper')
  }), /*#__PURE__*/React.createElement("span", {
    className: cssScope('__checked-ink')
  })), children ? /*#__PURE__*/React.createElement(Text, {
    variant: "plain",
    className: clsx(cssScope('__label-content'), disableState && cssScope('__label-content-disabled')),
    as: "span",
    lineClamp: isButton ? 1 : undefined,
    sx: sxLabelContent
  }, children) : null);
});
export default Radio;
