// @universe-design/react/src/avatar/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Basic from '../demo/0.basic';
import C1Type from '../demo/1.type';
import C2AutoFontSize from '../demo/2.auto-font-size';
import C3AvatarWithBadge from '../demo/3.avatar-with-badge';
import C4AvatarGroup from '../demo/4.avatar-group';
import C5AvatarGroupPopover from '../demo/5.avatar-group-popover';
if (!(C0Basic as Story).storyName) {
  (C0Basic as Story).storyName = '0. 基础用法';
  (C0Basic as Story).parameters = {
    docs: {
      description: {
        story: '简单的头像。头像有四种尺寸，两种形状可选。如果默认的尺寸无法满足需要，也可以使用数字尺寸。\n\n',
      },
    },
  };
}
if (!(C1Type as Story).storyName) {
  (C1Type as Story).storyName = '1. 头像类型';
  (C1Type as Story).parameters = {
    docs: {
      description: {
        story:
          '支持图片、图标和字符 3 种类型，其中图标和字符可以自定义图标颜色和背景。设置`colorStyle`为`outlined`可以得到空心头像。由于 UD 的设计中 shape 和是否有描边是一一对应的，所以为了避免业务使用不当进行，我们将该逻辑内置，`bordered` 属性将会被移除。\n\n',
      },
    },
  };
}
if (!(C2AutoFontSize as Story).storyName) {
  (C2AutoFontSize as Story).storyName = '2. 自动调整字体大小';
  (C2AutoFontSize as Story).parameters = {
    docs: {
      description: {
        story: '对于字符类型的头像，当字符串较长时，字体大小可以根据头像宽度自动调整。\n\n',
      },
    },
  };
}
if (!(C3AvatarWithBadge as Story).storyName) {
  (C3AvatarWithBadge as Story).storyName = '3. 带徽标头像';
  (C3AvatarWithBadge as Story).parameters = {
    docs: {
      description: {
        story: '通常用于消息提示。为了得到更好的视觉效果，我们会对内嵌元素的圆角进行计算，将徽标偏移到圆角中心。\n\n',
      },
    },
  };
}
if (!(C4AvatarGroup as Story).storyName) {
  (C4AvatarGroup as Story).storyName = '4. 头像群组';
  (C4AvatarGroup as Story).parameters = {
    docs: {
      description: {
        story: '使用`Avatar.Group`可以使用头像组功能，可通过`size`指定头像的大小。\n\n',
      },
    },
  };
}
if (!(C5AvatarGroupPopover as Story).storyName) {
  (C5AvatarGroupPopover as Story).storyName = '5. 自定义多余头像';
  (C5AvatarGroupPopover as Story).parameters = {
    docs: {
      description: {
        story:
          '可以使用 `maxProps` 覆盖多余头像的属性，也可以通过 `maxRenderer` 给多余头像添加气泡卡片或气泡文字。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Avatar/官网示例',
  tags: ['autodocs'],
};

export { C0Basic, C1Type, C2AutoFontSize, C3AvatarWithBadge, C4AvatarGroup, C5AvatarGroupPopover };
