// @universe-design/react/es/text/Text.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.uns1q8e{color:@udtoken-component-text-disabled-loading;}
import uzf2tb7 from "./../.stylex/uzf2tb7";
//.udrzuqr{cursor:not-allowed;}
import u1vjduc2 from "./../.stylex/u1vjduc2";
//.u1xqy1su{user-select:none;}
import u1443mj0 from "./../.stylex/u1443mj0";
//.u1j4dglz{font-style:italic;}
import u7llypb from "./../.stylex/u7llypb";
//.u1j59e10{text-align:justify;}
import uuojwkd from "./../.stylex/uuojwkd";
//.u12ymhq5{text-align:end;}
import u1d8k0i0 from "./../.stylex/u1d8k0i0";
//.u17mccla{text-align:center;}
import u1ust6hi from "./../.stylex/u1ust6hi";
//.upgzoln{text-align:start;}
import uuubajc from "./../.stylex/uuubajc";
//.u14t3ns0{display:inline-block;}
import u8wgnjj from "./../.stylex/u8wgnjj";
//.uhx3js9{word-wrap:break-word;}
import uitfv9r from "./../.stylex/uitfv9r";
//.uokg9q4{overflow-wrap:break-word;}
import u1pdmi33 from "./../.stylex/u1pdmi33";
/** @jsxRuntime classic */
/** @jsx jsx */
import * as React from 'react';
import stylex from '@universe-design/stylex';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx, useMergeComponentProps } from '@universe-design/stylex/react';
import { cssTypographyBody0, cssTypographyBody2, cssTypographyCaption0, cssTypographyCaption1, cssTypographyCaption1Hack, cssTypographyHeadline, cssTypographyTitle0, cssTypographyTitle1, cssTypographyTitle2, cssTypographyTitle3, cssTypographyTitle4, cssTypographyTitle5, sxLineClamp, sxOverflowEllipsis } from '../stylex';
import { genericForwardRef } from '../util/forwardRef';
import { useTextOverflowPopperSidecar } from '../config-provider/hooks/useTextOverflowPopperSidecar';
import { isHorizontalOverflowed, isVerticalOverflowed } from '../util/dom';
import { useCssScope } from '../config-provider';
import { isNonNullable } from '../util';
export const fontStyles = {
  title0: cssTypographyTitle0,
  title1: cssTypographyTitle1,
  title2: cssTypographyTitle2,
  title3: cssTypographyTitle3,
  title4: cssTypographyTitle4,
  title5: cssTypographyTitle5,
  headline: cssTypographyHeadline,
  caption0: cssTypographyCaption0,
  caption1: cssTypographyCaption1,
  caption1Hack: cssTypographyCaption1Hack,
  body0: cssTypographyBody0,
  body2: cssTypographyBody2
};

/**
 * @docgen
 */

export const Text = genericForwardRef((_ref, ref) => {
  let {
    as: C = 'span',
    variant = 'body0',
    lineClamp,
    disabled = false,
    tooltipProps,
    alignment,
    italic = false,
    ...props
  } = _ref;
  /**
   * ====================
   *   Props
   * ====================
   */
  const {
    children
  } = props;
  const noClamp = !isNonNullable(lineClamp);
  const oneLineClamp = !noClamp && lineClamp <= 1;
  const multilineClamp = !oneLineClamp && !noClamp;

  /**
   * ====================
   *   Context
   * ====================
   */
  const cssScope = useCssScope('text');
  const TextOverflowPopper = useTextOverflowPopperSidecar();

  /**
   * ====================
   *   Styles
   * ====================
   */
  const cProps = useMergeComponentProps(props, {
    className: cssScope(),
    sx: React.useMemo(() => [{
      "props": {
        "onut5h": uitfv9r,
        "Btd35i7": u1pdmi33
      }
    }, variant !== 'plain' && fontStyles[variant],
    // 如果 span 并且开了 one line clamp 则需要使用 inline-block
    C === 'span' && oneLineClamp && {
      "props": {
        "mc9l5x": u8wgnjj
      }
    },
    // line clamp 里面有 display: -webkit-box 要放在其他的 display 后面
    (() => {
      if (oneLineClamp) {
        return sxOverflowEllipsis;
      }
      if (multilineClamp) {
        return sxLineClamp;
      }
    })(),
    // 默认从父级继承或使用 initial 值
    alignment && {
      start: {
        "props": {
          "fsow6f": uuubajc
        }
      },
      center: {
        "props": {
          "fsow6f": u1ust6hi
        }
      },
      end: {
        "props": {
          "fsow6f": u1d8k0i0
        }
      },
      justify: {
        "props": {
          "fsow6f": uuojwkd
        }
      }
    }[alignment], italic && {
      "props": {
        "B80ckks": u7llypb
      }
    }, disabled && {
      "props": {
        "sj55zd": uzf2tb7,
        "Bceei9c": u1vjduc2,
        "famaaq": u1443mj0
      }
    }], [C, alignment, disabled, italic, multilineClamp, oneLineClamp, variant]),
    style: {
      WebkitLineClamp: multilineClamp ? lineClamp : undefined
    }
  });

  /**
   * ====================
   *   Render
   * ====================
   */
  const node = jsx(C, _extends({
    ref: ref
  }, cProps));
  if (noClamp) {
    return node;
  }
  const checkOverflow = oneLineClamp ? isHorizontalOverflowed : isVerticalOverflowed;
  return jsx(TextOverflowPopper, _extends({
    title: children,
    checkOverflow
  }, tooltipProps), node);
});

/**
 * @docgen
 */

if (process.env.NODE_ENV !== "production") {
  Text.displayName = 'Text';
}
