// @universe-design/styles/cjs/calendar/index.dark.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.caltoken_event_dark_bg_tur = exports.caltoken_event_dark_bg_red = exports.caltoken_event_dark_bg_purple = exports.caltoken_event_dark_bg_pending_yellow = exports.caltoken_event_dark_bg_pending_wathet = exports.caltoken_event_dark_bg_pending_violet = exports.caltoken_event_dark_bg_pending_tur = exports.caltoken_event_dark_bg_pending_red = exports.caltoken_event_dark_bg_pending_purple = exports.caltoken_event_dark_bg_pending_orange = exports.caltoken_event_dark_bg_pending_neutral = exports.caltoken_event_dark_bg_pending_indigo = exports.caltoken_event_dark_bg_pending_green = exports.caltoken_event_dark_bg_pending_carmine = exports.caltoken_event_dark_bg_pending_blue = exports.caltoken_event_dark_bg_orange = exports.caltoken_event_dark_bg_neutral = exports.caltoken_event_dark_bg_indigo = exports.caltoken_event_dark_bg_green = exports.caltoken_event_dark_bg_carmine = exports.caltoken_event_dark_bg_blue = exports.caltoken_event_dark_bg_bar_yellow = exports.caltoken_event_dark_bg_bar_wathet = exports.caltoken_event_dark_bg_bar_violet = exports.caltoken_event_dark_bg_bar_tur = exports.caltoken_event_dark_bg_bar_red = exports.caltoken_event_dark_bg_bar_purple = exports.caltoken_event_dark_bg_bar_orange = exports.caltoken_event_dark_bg_bar_neutral = exports.caltoken_event_dark_bg_bar_indigo = exports.caltoken_event_dark_bg_bar_green = exports.caltoken_event_dark_bg_bar_carmine = exports.caltoken_event_dark_bg_bar_blue = exports.caltoken_event_conflict_event_shadow = exports.caltoken_event_border_unanswered_yellow = exports.caltoken_event_border_unanswered_wathet = exports.caltoken_event_border_unanswered_violet = exports.caltoken_event_border_unanswered_tur = exports.caltoken_event_border_unanswered_red = exports.caltoken_event_border_unanswered_purple = exports.caltoken_event_border_unanswered_orange = exports.caltoken_event_border_unanswered_neutral = exports.caltoken_event_border_unanswered_indigo = exports.caltoken_event_border_unanswered_green = exports.caltoken_event_border_unanswered_carmine = exports.caltoken_event_border_unanswered_blue = exports.caltoken_btn_selected_line_gray = exports.caltoken_addroms_bg_icon_3 = exports.caltoken_addroms_bg_icon_2 = exports.caltoken_addroms_bg_icon_1 = void 0;
exports.caltoken_event_light_bg_carmine = exports.caltoken_event_light_bg_blue = exports.caltoken_event_light_bg_bar_yellow = exports.caltoken_event_light_bg_bar_wathet = exports.caltoken_event_light_bg_bar_violet = exports.caltoken_event_light_bg_bar_tur = exports.caltoken_event_light_bg_bar_red = exports.caltoken_event_light_bg_bar_purple = exports.caltoken_event_light_bg_bar_orange = exports.caltoken_event_light_bg_bar_neutral = exports.caltoken_event_light_bg_bar_indigo = exports.caltoken_event_light_bg_bar_green = exports.caltoken_event_light_bg_bar_carmine = exports.caltoken_event_light_bg_bar_blue = exports.caltoken_event_detail_bg_right_wathet = exports.caltoken_event_detail_bg_right_tur = exports.caltoken_event_detail_bg_right_red = exports.caltoken_event_detail_bg_right_orange = exports.caltoken_event_detail_bg_right_green = exports.caltoken_event_detail_bg_left_wathet = exports.caltoken_event_detail_bg_left_tur = exports.caltoken_event_detail_bg_left_red = exports.caltoken_event_detail_bg_left_orange = exports.caltoken_event_detail_bg_left_green = exports.caltoken_event_dark_pending_bar_yellow = exports.caltoken_event_dark_pending_bar_wathet = exports.caltoken_event_dark_pending_bar_violet = exports.caltoken_event_dark_pending_bar_tur = exports.caltoken_event_dark_pending_bar_red = exports.caltoken_event_dark_pending_bar_purple = exports.caltoken_event_dark_pending_bar_orange = exports.caltoken_event_dark_pending_bar_neutral = exports.caltoken_event_dark_pending_bar_indigo = exports.caltoken_event_dark_pending_bar_green = exports.caltoken_event_dark_pending_bar_carmine = exports.caltoken_event_dark_pending_bar_blue = exports.caltoken_event_dark_calendar_yellow = exports.caltoken_event_dark_calendar_wathet = exports.caltoken_event_dark_calendar_violet = exports.caltoken_event_dark_calendar_tur = exports.caltoken_event_dark_calendar_red = exports.caltoken_event_dark_calendar_purple = exports.caltoken_event_dark_calendar_orange = exports.caltoken_event_dark_calendar_neutral = exports.caltoken_event_dark_calendar_indigo = exports.caltoken_event_dark_calendar_green = exports.caltoken_event_dark_calendar_carmine = exports.caltoken_event_dark_bg_yellow = exports.caltoken_event_dark_bg_wathet = exports.caltoken_event_dark_bg_violet = void 0;
exports.caltoken_event_light_text_unanswered_purple = exports.caltoken_event_light_text_unanswered_orange = exports.caltoken_event_light_text_unanswered_neutral = exports.caltoken_event_light_text_unanswered_indigo = exports.caltoken_event_light_text_unanswered_green = exports.caltoken_event_light_text_unanswered_carmine = exports.caltoken_event_light_text_unanswered_blue = exports.caltoken_event_light_text_tur = exports.caltoken_event_light_text_red = exports.caltoken_event_light_text_purple = exports.caltoken_event_light_text_orange = exports.caltoken_event_light_text_neutral = exports.caltoken_event_light_text_indigo = exports.caltoken_event_light_text_green = exports.caltoken_event_light_text_carmine = exports.caltoken_event_light_text_blue = exports.caltoken_event_light_calendar_yellow = exports.caltoken_event_light_calendar_wathet = exports.caltoken_event_light_calendar_violet = exports.caltoken_event_light_calendar_tur = exports.caltoken_event_light_calendar_red = exports.caltoken_event_light_calendar_purple = exports.caltoken_event_light_calendar_orange = exports.caltoken_event_light_calendar_neutral = exports.caltoken_event_light_calendar_indigo = exports.caltoken_event_light_calendar_green = exports.caltoken_event_light_calendar_carmine = exports.caltoken_event_light_calendar_blue = exports.caltoken_event_light_bg_yellow = exports.caltoken_event_light_bg_wathet = exports.caltoken_event_light_bg_violet = exports.caltoken_event_light_bg_tur = exports.caltoken_event_light_bg_red = exports.caltoken_event_light_bg_purple = exports.caltoken_event_light_bg_pending_yellow = exports.caltoken_event_light_bg_pending_wathet = exports.caltoken_event_light_bg_pending_violet = exports.caltoken_event_light_bg_pending_tur = exports.caltoken_event_light_bg_pending_red = exports.caltoken_event_light_bg_pending_purple = exports.caltoken_event_light_bg_pending_orange = exports.caltoken_event_light_bg_pending_neutral = exports.caltoken_event_light_bg_pending_indigo = exports.caltoken_event_light_bg_pending_green = exports.caltoken_event_light_bg_pending_carmine = exports.caltoken_event_light_bg_pending_blue = exports.caltoken_event_light_bg_oran = exports.caltoken_event_light_bg_neutral = exports.caltoken_event_light_bg_indigo = exports.caltoken_event_light_bg_green = void 0;
exports.caltoken_tag_colour_day = exports.caltoken_event_stripe_yellow = exports.caltoken_event_stripe_wathet = exports.caltoken_event_stripe_violet = exports.caltoken_event_stripe_tur = exports.caltoken_event_stripe_red = exports.caltoken_event_stripe_purple = exports.caltoken_event_stripe_orange = exports.caltoken_event_stripe_neutral = exports.caltoken_event_stripe_indigo = exports.caltoken_event_stripe_green = exports.caltoken_event_stripe_carmine = exports.caltoken_event_stripe_blue = exports.caltoken_event_quick_creation_shadow = exports.caltoken_event_light_text_yellow = exports.caltoken_event_light_text_wathet = exports.caltoken_event_light_text_violet = exports.caltoken_event_light_text_unanswered_yellow = exports.caltoken_event_light_text_unanswered_wathet = exports.caltoken_event_light_text_unanswered_violet = exports.caltoken_event_light_text_unanswered_tur = exports.caltoken_event_light_text_unanswered_red = void 0;
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
exports.caltoken_addroms_bg_icon_1 = '#173b12';
exports.caltoken_addroms_bg_icon_2 = '#21511a';
exports.caltoken_addroms_bg_icon_3 = '#51ba43';
exports.caltoken_btn_selected_line_gray = '#a6a6a6';
exports.caltoken_event_border_unanswered_blue = '#5a7aba';
exports.caltoken_event_border_unanswered_carmine = '#c24a8e';
exports.caltoken_event_border_unanswered_green = '#2f7526';
exports.caltoken_event_border_unanswered_indigo = '#5e68e8';
exports.caltoken_event_border_unanswered_neutral = '#757575';
exports.caltoken_event_border_unanswered_orange = '#b85e1a';
exports.caltoken_event_border_unanswered_purple = '#8c55ec';
exports.caltoken_event_border_unanswered_red = '#d14642';
exports.caltoken_event_border_unanswered_tur = '#227369';
exports.caltoken_event_border_unanswered_violet = '#a43da4';
exports.caltoken_event_border_unanswered_wathet = '#1099cc';
exports.caltoken_event_border_unanswered_yellow = '#ad7d15';
exports.caltoken_event_conflict_event_shadow = '8px 8px 24px 8px rgba(0, 0, 0, 0.11), -6px 6px 12px rgba(0, 0, 0, 0.05), -4px 4px 8px -8px rgba(0, 0, 0, 0.08)';
exports.caltoken_event_dark_bg_bar_blue = '#4c88ff';
exports.caltoken_event_dark_bg_bar_carmine = '#db5ea4';
exports.caltoken_event_dark_bg_bar_green = '#35872a';
exports.caltoken_event_dark_bg_bar_indigo = '#7b83f7';
exports.caltoken_event_dark_bg_bar_neutral = '#757575';
exports.caltoken_event_dark_bg_bar_orange = '#db7018';
exports.caltoken_event_dark_bg_bar_purple = '#a575fa';
exports.caltoken_event_dark_bg_bar_red = '#f05b56';
exports.caltoken_event_dark_bg_bar_tur = '#1fa18f';
exports.caltoken_event_dark_bg_bar_violet = '#d661d6';
exports.caltoken_event_dark_bg_bar_wathet = '#1099cc';
exports.caltoken_event_dark_bg_bar_yellow = '#d49b0b';
exports.caltoken_event_dark_bg_blue = '#3370eb';
exports.caltoken_event_dark_bg_carmine = '#c24a8e';
exports.caltoken_event_dark_bg_green = '#2f7526';
exports.caltoken_event_dark_bg_indigo = '#5e68e8';
exports.caltoken_event_dark_bg_neutral = '#5f5f5f';
exports.caltoken_event_dark_bg_orange = '#b85e1a';
exports.caltoken_event_dark_bg_pending_blue = '#275fce';
exports.caltoken_event_dark_bg_pending_carmine = '#ab417d';
exports.caltoken_event_dark_bg_pending_green = '#296621';
exports.caltoken_event_dark_bg_pending_indigo = '#515ad6';
exports.caltoken_event_dark_bg_pending_neutral = '#505050';
exports.caltoken_event_dark_bg_pending_orange = '#a15317';
exports.caltoken_event_dark_bg_pending_purple = '#7c4ad4';
exports.caltoken_event_dark_bg_pending_red = '#b33a37';
exports.caltoken_event_dark_bg_pending_tur = '#227369';
exports.caltoken_event_dark_bg_pending_violet = '#a43da4';
exports.caltoken_event_dark_bg_pending_wathet = '#146c91';
exports.caltoken_event_dark_bg_pending_yellow = '#8a6419';
exports.caltoken_event_dark_bg_purple = '#8c55ec';
exports.caltoken_event_dark_bg_red = '#d14642';
exports.caltoken_event_dark_bg_tur = '#198578';
exports.caltoken_event_dark_bg_violet = '#b54ab5';
exports.caltoken_event_dark_bg_wathet = '#1a7fab';
exports.caltoken_event_dark_bg_yellow = '#ad7d15';
exports.caltoken_event_dark_calendar_carmine = '#f01d94';
exports.caltoken_event_dark_calendar_green = '#2ea121';
exports.caltoken_event_dark_calendar_indigo = '#4954e6';
exports.caltoken_event_dark_calendar_neutral = '#646a73';
exports.caltoken_event_dark_calendar_orange = '#de7802';
exports.caltoken_event_dark_calendar_purple = '#7f3bf5';
exports.caltoken_event_dark_calendar_red = '#f54a45';
exports.caltoken_event_dark_calendar_tur = '#04b49c';
exports.caltoken_event_dark_calendar_violet = '#b320b3';
exports.caltoken_event_dark_calendar_wathet = '#049fd7';
exports.caltoken_event_dark_calendar_yellow = '#dc9b04';
exports.caltoken_event_dark_pending_bar_blue = '#4c88ff';
exports.caltoken_event_dark_pending_bar_carmine = '#db5ea4';
exports.caltoken_event_dark_pending_bar_green = '#419e34';
exports.caltoken_event_dark_pending_bar_indigo = '#7b83f7';
exports.caltoken_event_dark_pending_bar_neutral = '#757575';
exports.caltoken_event_dark_pending_bar_orange = '#db7018';
exports.caltoken_event_dark_pending_bar_purple = '#a575fa';
exports.caltoken_event_dark_pending_bar_red = '#f05b56';
exports.caltoken_event_dark_pending_bar_tur = '#1fa18f';
exports.caltoken_event_dark_pending_bar_violet = '#d661d6';
exports.caltoken_event_dark_pending_bar_wathet = '#1099cc';
exports.caltoken_event_dark_pending_bar_yellow = '#d49b0b';
exports.caltoken_event_detail_bg_left_green = '#419e34';
exports.caltoken_event_detail_bg_left_orange = '#db7018';
exports.caltoken_event_detail_bg_left_red = '#f05b56';
exports.caltoken_event_detail_bg_left_tur = '#1fa18f';
exports.caltoken_event_detail_bg_left_wathet = '#1a7fab';
exports.caltoken_event_detail_bg_right_green = '#51ba43';
exports.caltoken_event_detail_bg_right_orange = '#f3871b';
exports.caltoken_event_detail_bg_right_red = '#f05b56';
exports.caltoken_event_detail_bg_right_tur = '#1ab7a1';
exports.caltoken_event_detail_bg_right_wathet = '#1099cc';
exports.caltoken_event_light_bg_bar_blue = '#4c88ff';
exports.caltoken_event_light_bg_bar_carmine = '#db5ea4';
exports.caltoken_event_light_bg_bar_green = '#419e34';
exports.caltoken_event_light_bg_bar_indigo = '#7b83f7';
exports.caltoken_event_light_bg_bar_neutral = '#757575';
exports.caltoken_event_light_bg_bar_orange = '#b85e1a';
exports.caltoken_event_light_bg_bar_purple = '#a575fa';
exports.caltoken_event_light_bg_bar_red = '#d14642';
exports.caltoken_event_light_bg_bar_tur = '#1fa18f';
exports.caltoken_event_light_bg_bar_violet = '#d661d6';
exports.caltoken_event_light_bg_bar_wathet = '#1099cc';
exports.caltoken_event_light_bg_bar_yellow = '#d49b0b';
exports.caltoken_event_light_bg_blue = '#173166';
exports.caltoken_event_light_bg_carmine = '#591c3f';
exports.caltoken_event_light_bg_green = '#173b12';
exports.caltoken_event_light_bg_indigo = '#2a2d69';
exports.caltoken_event_light_bg_neutral = '#373737';
exports.caltoken_event_light_bg_oran = '#4a2b10';
exports.caltoken_event_light_bg_pending_blue = '#152340';
exports.caltoken_event_light_bg_pending_carmine = '#3a182b';
exports.caltoken_event_light_bg_pending_green = '#0e2b0a';
exports.caltoken_event_light_bg_pending_indigo = '#1e204a';
exports.caltoken_event_light_bg_pending_neutral = '#292929';
exports.caltoken_event_light_bg_pending_orange = '#33210b';
exports.caltoken_event_light_bg_pending_purple = '#2b194a';
exports.caltoken_event_light_bg_pending_red = '#3d1a19';
exports.caltoken_event_light_bg_pending_tur = '#132926';
exports.caltoken_event_light_bg_pending_violet = '#3b153b';
exports.caltoken_event_light_bg_pending_wathet = '#152830';
exports.caltoken_event_light_bg_pending_yellow = '#30250a';
exports.caltoken_event_light_bg_purple = '#3f2073';
exports.caltoken_event_light_bg_red = '#591f1d';
exports.caltoken_event_light_bg_tur = '#173b36';
exports.caltoken_event_light_bg_violet = '#541854';
exports.caltoken_event_light_bg_wathet = '#103647';
exports.caltoken_event_light_bg_yellow = '#473409';
exports.caltoken_event_light_calendar_blue = '#4e83fd';
exports.caltoken_event_light_calendar_carmine = '#f14ba9';
exports.caltoken_event_light_calendar_green = '#34c724';
exports.caltoken_event_light_calendar_indigo = '#616ae5';
exports.caltoken_event_light_calendar_neutral = '#9ca2a9';
exports.caltoken_event_light_calendar_orange = '#ff8800';
exports.caltoken_event_light_calendar_purple = '#935af6';
exports.caltoken_event_light_calendar_red = '#f76964';
exports.caltoken_event_light_calendar_tur = '#00d6b9';
exports.caltoken_event_light_calendar_violet = '#d136d1';
exports.caltoken_event_light_calendar_wathet = '#14c0ff';
exports.caltoken_event_light_calendar_yellow = '#ffc60a';
exports.caltoken_event_light_text_blue = '#75a4ff';
exports.caltoken_event_light_text_carmine = '#ed77ba';
exports.caltoken_event_light_text_green = '#51ba43';
exports.caltoken_event_light_text_indigo = '#9499f7';
exports.caltoken_event_light_text_neutral = '#cfcfcf';
exports.caltoken_event_light_text_orange = '#f89e44';
exports.caltoken_event_light_text_purple = '#b88ffe';
exports.caltoken_event_light_text_red = '#f6827e';
exports.caltoken_event_light_text_tur = '#1ab7a1';
exports.caltoken_event_light_text_unanswered_blue = '#6c96e8';
exports.caltoken_event_light_text_unanswered_carmine = '#db5ea4';
exports.caltoken_event_light_text_unanswered_green = '#419e34';
exports.caltoken_event_light_text_unanswered_indigo = '#7b83f7';
exports.caltoken_event_light_text_unanswered_neutral = '#a6a6a6';
exports.caltoken_event_light_text_unanswered_orange = '#e08e41';
exports.caltoken_event_light_text_unanswered_purple = '#a575fa';
exports.caltoken_event_light_text_unanswered_red = '#dc7572';
exports.caltoken_event_light_text_unanswered_tur = '#1fa18f';
exports.caltoken_event_light_text_unanswered_violet = '#cd75cd';
exports.caltoken_event_light_text_unanswered_wathet = '#1099cc';
exports.caltoken_event_light_text_unanswered_yellow = '#d49b0b';
exports.caltoken_event_light_text_violet = '#e17fe1';
exports.caltoken_event_light_text_wathet = '#25b2e5';
exports.caltoken_event_light_text_yellow = '#f0b622';
exports.caltoken_event_quick_creation_shadow = '0px 10px 36px 10px rgba(0, 0, 0, 0.09), 0px 8px 24px rgba(0, 0, 0, 0.09), 0px 6px 12px -10px rgba(0, 0, 0, 0.11)';
exports.caltoken_event_stripe_blue = '#152340';
exports.caltoken_event_stripe_carmine = '#3a182b';
exports.caltoken_event_stripe_green = '#0e2b0a';
exports.caltoken_event_stripe_indigo = '#1e204a';
exports.caltoken_event_stripe_neutral = '#292929';
exports.caltoken_event_stripe_orange = '#33210b';
exports.caltoken_event_stripe_purple = '#2b194a';
exports.caltoken_event_stripe_red = '#3d1a19';
exports.caltoken_event_stripe_tur = '#132926';
exports.caltoken_event_stripe_violet = '#3b153b';
exports.caltoken_event_stripe_wathet = '#152830';
exports.caltoken_event_stripe_yellow = '#30250a';
exports.caltoken_tag_colour_day = '#cf9500';
