// @universe-design/styles/cjs/calendar/index.light.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.caltoken_event_dark_bg_tur = exports.caltoken_event_dark_bg_red = exports.caltoken_event_dark_bg_purple = exports.caltoken_event_dark_bg_pending_yellow = exports.caltoken_event_dark_bg_pending_wathet = exports.caltoken_event_dark_bg_pending_violet = exports.caltoken_event_dark_bg_pending_tur = exports.caltoken_event_dark_bg_pending_red = exports.caltoken_event_dark_bg_pending_purple = exports.caltoken_event_dark_bg_pending_orange = exports.caltoken_event_dark_bg_pending_neutral = exports.caltoken_event_dark_bg_pending_indigo = exports.caltoken_event_dark_bg_pending_green = exports.caltoken_event_dark_bg_pending_carmine = exports.caltoken_event_dark_bg_pending_blue = exports.caltoken_event_dark_bg_orange = exports.caltoken_event_dark_bg_neutral = exports.caltoken_event_dark_bg_indigo = exports.caltoken_event_dark_bg_green = exports.caltoken_event_dark_bg_carmine = exports.caltoken_event_dark_bg_blue = exports.caltoken_event_dark_bg_bar_yellow = exports.caltoken_event_dark_bg_bar_wathet = exports.caltoken_event_dark_bg_bar_violet = exports.caltoken_event_dark_bg_bar_tur = exports.caltoken_event_dark_bg_bar_red = exports.caltoken_event_dark_bg_bar_purple = exports.caltoken_event_dark_bg_bar_orange = exports.caltoken_event_dark_bg_bar_neutral = exports.caltoken_event_dark_bg_bar_indigo = exports.caltoken_event_dark_bg_bar_green = exports.caltoken_event_dark_bg_bar_carmine = exports.caltoken_event_dark_bg_bar_blue = exports.caltoken_event_conflict_event_shadow = exports.caltoken_event_border_unanswered_yellow = exports.caltoken_event_border_unanswered_wathet = exports.caltoken_event_border_unanswered_violet = exports.caltoken_event_border_unanswered_tur = exports.caltoken_event_border_unanswered_red = exports.caltoken_event_border_unanswered_purple = exports.caltoken_event_border_unanswered_orange = exports.caltoken_event_border_unanswered_neutral = exports.caltoken_event_border_unanswered_indigo = exports.caltoken_event_border_unanswered_green = exports.caltoken_event_border_unanswered_carmine = exports.caltoken_event_border_unanswered_blue = exports.caltoken_btn_selected_line_gray = exports.caltoken_addroms_bg_icon_3 = exports.caltoken_addroms_bg_icon_2 = exports.caltoken_addroms_bg_icon_1 = void 0;
exports.caltoken_event_light_bg_carmine = exports.caltoken_event_light_bg_blue = exports.caltoken_event_light_bg_bar_yellow = exports.caltoken_event_light_bg_bar_wathet = exports.caltoken_event_light_bg_bar_violet = exports.caltoken_event_light_bg_bar_tur = exports.caltoken_event_light_bg_bar_red = exports.caltoken_event_light_bg_bar_purple = exports.caltoken_event_light_bg_bar_orange = exports.caltoken_event_light_bg_bar_neutral = exports.caltoken_event_light_bg_bar_indigo = exports.caltoken_event_light_bg_bar_green = exports.caltoken_event_light_bg_bar_carmine = exports.caltoken_event_light_bg_bar_blue = exports.caltoken_event_detail_bg_right_wathet = exports.caltoken_event_detail_bg_right_tur = exports.caltoken_event_detail_bg_right_red = exports.caltoken_event_detail_bg_right_orange = exports.caltoken_event_detail_bg_right_green = exports.caltoken_event_detail_bg_left_wathet = exports.caltoken_event_detail_bg_left_tur = exports.caltoken_event_detail_bg_left_red = exports.caltoken_event_detail_bg_left_orange = exports.caltoken_event_detail_bg_left_green = exports.caltoken_event_dark_pending_bar_yellow = exports.caltoken_event_dark_pending_bar_wathet = exports.caltoken_event_dark_pending_bar_violet = exports.caltoken_event_dark_pending_bar_tur = exports.caltoken_event_dark_pending_bar_red = exports.caltoken_event_dark_pending_bar_purple = exports.caltoken_event_dark_pending_bar_orange = exports.caltoken_event_dark_pending_bar_neutral = exports.caltoken_event_dark_pending_bar_indigo = exports.caltoken_event_dark_pending_bar_green = exports.caltoken_event_dark_pending_bar_carmine = exports.caltoken_event_dark_pending_bar_blue = exports.caltoken_event_dark_calendar_yellow = exports.caltoken_event_dark_calendar_wathet = exports.caltoken_event_dark_calendar_violet = exports.caltoken_event_dark_calendar_tur = exports.caltoken_event_dark_calendar_red = exports.caltoken_event_dark_calendar_purple = exports.caltoken_event_dark_calendar_orange = exports.caltoken_event_dark_calendar_neutral = exports.caltoken_event_dark_calendar_indigo = exports.caltoken_event_dark_calendar_green = exports.caltoken_event_dark_calendar_carmine = exports.caltoken_event_dark_bg_yellow = exports.caltoken_event_dark_bg_wathet = exports.caltoken_event_dark_bg_violet = void 0;
exports.caltoken_event_light_text_unanswered_purple = exports.caltoken_event_light_text_unanswered_orange = exports.caltoken_event_light_text_unanswered_neutral = exports.caltoken_event_light_text_unanswered_indigo = exports.caltoken_event_light_text_unanswered_green = exports.caltoken_event_light_text_unanswered_carmine = exports.caltoken_event_light_text_unanswered_blue = exports.caltoken_event_light_text_tur = exports.caltoken_event_light_text_red = exports.caltoken_event_light_text_purple = exports.caltoken_event_light_text_orange = exports.caltoken_event_light_text_neutral = exports.caltoken_event_light_text_indigo = exports.caltoken_event_light_text_green = exports.caltoken_event_light_text_carmine = exports.caltoken_event_light_text_blue = exports.caltoken_event_light_calendar_yellow = exports.caltoken_event_light_calendar_wathet = exports.caltoken_event_light_calendar_violet = exports.caltoken_event_light_calendar_tur = exports.caltoken_event_light_calendar_red = exports.caltoken_event_light_calendar_purple = exports.caltoken_event_light_calendar_orange = exports.caltoken_event_light_calendar_neutral = exports.caltoken_event_light_calendar_indigo = exports.caltoken_event_light_calendar_green = exports.caltoken_event_light_calendar_carmine = exports.caltoken_event_light_calendar_blue = exports.caltoken_event_light_bg_yellow = exports.caltoken_event_light_bg_wathet = exports.caltoken_event_light_bg_violet = exports.caltoken_event_light_bg_tur = exports.caltoken_event_light_bg_red = exports.caltoken_event_light_bg_purple = exports.caltoken_event_light_bg_pending_yellow = exports.caltoken_event_light_bg_pending_wathet = exports.caltoken_event_light_bg_pending_violet = exports.caltoken_event_light_bg_pending_tur = exports.caltoken_event_light_bg_pending_red = exports.caltoken_event_light_bg_pending_purple = exports.caltoken_event_light_bg_pending_orange = exports.caltoken_event_light_bg_pending_neutral = exports.caltoken_event_light_bg_pending_indigo = exports.caltoken_event_light_bg_pending_green = exports.caltoken_event_light_bg_pending_carmine = exports.caltoken_event_light_bg_pending_blue = exports.caltoken_event_light_bg_oran = exports.caltoken_event_light_bg_neutral = exports.caltoken_event_light_bg_indigo = exports.caltoken_event_light_bg_green = void 0;
exports.caltoken_tag_colour_day = exports.caltoken_event_stripe_yellow = exports.caltoken_event_stripe_wathet = exports.caltoken_event_stripe_violet = exports.caltoken_event_stripe_tur = exports.caltoken_event_stripe_red = exports.caltoken_event_stripe_purple = exports.caltoken_event_stripe_orange = exports.caltoken_event_stripe_neutral = exports.caltoken_event_stripe_indigo = exports.caltoken_event_stripe_green = exports.caltoken_event_stripe_carmine = exports.caltoken_event_stripe_blue = exports.caltoken_event_quick_creation_shadow = exports.caltoken_event_light_text_yellow = exports.caltoken_event_light_text_wathet = exports.caltoken_event_light_text_violet = exports.caltoken_event_light_text_unanswered_yellow = exports.caltoken_event_light_text_unanswered_wathet = exports.caltoken_event_light_text_unanswered_violet = exports.caltoken_event_light_text_unanswered_tur = exports.caltoken_event_light_text_unanswered_red = void 0;
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
exports.caltoken_addroms_bg_icon_1 = '#e4fae1';
exports.caltoken_addroms_bg_icon_2 = '#d0f5ce';
exports.caltoken_addroms_bg_icon_3 = '#5cd168';
exports.caltoken_btn_selected_line_gray = '#646a73';
exports.caltoken_event_border_unanswered_blue = '#336df4';
exports.caltoken_event_border_unanswered_carmine = '#df58a5';
exports.caltoken_event_border_unanswered_green = '#32a645';
exports.caltoken_event_border_unanswered_indigo = '#757df0';
exports.caltoken_event_border_unanswered_neutral = '#8f959e';
exports.caltoken_event_border_unanswered_orange = '#ed6d0c';
exports.caltoken_event_border_unanswered_purple = '#9f6ff1';
exports.caltoken_event_border_unanswered_red = '#f54a45';
exports.caltoken_event_border_unanswered_tur = '#10a893';
exports.caltoken_event_border_unanswered_violet = '#cf5ecf';
exports.caltoken_event_border_unanswered_wathet = '#1295ca';
exports.caltoken_event_border_unanswered_yellow = '#d99904';
exports.caltoken_event_conflict_event_shadow = '8px 8px 24px 8px rgba(31, 35, 41, 0.11), -6px 6px 12px rgba(31, 35, 41, 0.05), -4px 4px 8px -8px rgba(31, 35, 41, 0.08)';
exports.caltoken_event_dark_bg_bar_blue = '#336df4';
exports.caltoken_event_dark_bg_bar_carmine = '#cc398c';
exports.caltoken_event_dark_bg_bar_green = '#258832';
exports.caltoken_event_dark_bg_bar_indigo = '#5b65f5';
exports.caltoken_event_dark_bg_bar_neutral = 'rgba(100, 106, 115, 0.6)';
exports.caltoken_event_dark_bg_bar_orange = '#c25705';
exports.caltoken_event_dark_bg_bar_purple = '#8d55ed';
exports.caltoken_event_dark_bg_bar_red = '#e22e28';
exports.caltoken_event_dark_bg_bar_tur = '#0f8575';
exports.caltoken_event_dark_bg_bar_violet = '#bf3dbf';
exports.caltoken_event_dark_bg_bar_wathet = '#047fb0';
exports.caltoken_event_dark_bg_bar_yellow = '#865b03';
exports.caltoken_event_dark_bg_blue = '#5083fb';
exports.caltoken_event_dark_bg_carmine = '#df58a5';
exports.caltoken_event_dark_bg_green = '#32a645';
exports.caltoken_event_dark_bg_indigo = '#757df0';
exports.caltoken_event_dark_bg_neutral = '#83888f';
exports.caltoken_event_dark_bg_orange = '#ed6d0c';
exports.caltoken_event_dark_bg_pending_blue = '#5083fb';
exports.caltoken_event_dark_bg_pending_carmine = '#e369ae';
exports.caltoken_event_dark_bg_pending_green = '#46a73b';
exports.caltoken_event_dark_bg_pending_indigo = '#757df0';
exports.caltoken_event_dark_bg_pending_neutral = '#8f959e';
exports.caltoken_event_dark_bg_pending_orange = '#ed6d0c';
exports.caltoken_event_dark_bg_pending_purple = '#9f6ff1';
exports.caltoken_event_dark_bg_pending_red = '#f5605b';
exports.caltoken_event_dark_bg_pending_tur = '#10a893';
exports.caltoken_event_dark_bg_pending_violet = '#cf5ecf';
exports.caltoken_event_dark_bg_pending_wathet = '#1295ca';
exports.caltoken_event_dark_bg_pending_yellow = '#ad7a03';
exports.caltoken_event_dark_bg_purple = '#9f6ff1';
exports.caltoken_event_dark_bg_red = '#f5605b';
exports.caltoken_event_dark_bg_tur = '#10a893';
exports.caltoken_event_dark_bg_violet = '#cf5ecf';
exports.caltoken_event_dark_bg_wathet = '#1295ca';
exports.caltoken_event_dark_bg_yellow = '#ad7a03';
exports.caltoken_event_dark_calendar_carmine = '#f01d94';
exports.caltoken_event_dark_calendar_green = '#2ea121';
exports.caltoken_event_dark_calendar_indigo = '#4954e6';
exports.caltoken_event_dark_calendar_neutral = '#646a73';
exports.caltoken_event_dark_calendar_orange = '#de7802';
exports.caltoken_event_dark_calendar_purple = '#7f3bf5';
exports.caltoken_event_dark_calendar_red = '#f54a45';
exports.caltoken_event_dark_calendar_tur = '#04b49c';
exports.caltoken_event_dark_calendar_violet = '#b320b3';
exports.caltoken_event_dark_calendar_wathet = '#049fd7';
exports.caltoken_event_dark_calendar_yellow = '#dc9b04';
exports.caltoken_event_dark_pending_bar_blue = '#336df4';
exports.caltoken_event_dark_pending_bar_carmine = '#cc398c';
exports.caltoken_event_dark_pending_bar_green = '#1a7526';
exports.caltoken_event_dark_pending_bar_indigo = '#5b65f5';
exports.caltoken_event_dark_pending_bar_neutral = 'rgba(100, 106, 115, 0.85)';
exports.caltoken_event_dark_pending_bar_orange = '#c25705';
exports.caltoken_event_dark_pending_bar_purple = '#8d55ed';
exports.caltoken_event_dark_pending_bar_red = '#e22e28';
exports.caltoken_event_dark_pending_bar_tur = '#0f8575';
exports.caltoken_event_dark_pending_bar_violet = '#bf3dbf';
exports.caltoken_event_dark_pending_bar_wathet = '#047fb0';
exports.caltoken_event_dark_pending_bar_yellow = '#865b03';
exports.caltoken_event_detail_bg_left_green = '#5cd168';
exports.caltoken_event_detail_bg_left_orange = '#ff9d4c';
exports.caltoken_event_detail_bg_left_red = '#ff7570';
exports.caltoken_event_detail_bg_left_tur = '#2dbeab';
exports.caltoken_event_detail_bg_left_wathet = '#25b0e7';
exports.caltoken_event_detail_bg_right_green = '#32a645';
exports.caltoken_event_detail_bg_right_orange = '#ed6d0c';
exports.caltoken_event_detail_bg_right_red = '#f54a45';
exports.caltoken_event_detail_bg_right_tur = '#10a893';
exports.caltoken_event_detail_bg_right_wathet = '#1295ca';
exports.caltoken_event_light_bg_bar_blue = '#5083fb';
exports.caltoken_event_light_bg_bar_carmine = '#df58a5';
exports.caltoken_event_light_bg_bar_green = '#32a645';
exports.caltoken_event_light_bg_bar_indigo = '#757df0';
exports.caltoken_event_light_bg_bar_neutral = '#8f959e';
exports.caltoken_event_light_bg_bar_orange = '#ed6d0c';
exports.caltoken_event_light_bg_bar_purple = '#9f6ff1';
exports.caltoken_event_light_bg_bar_red = '#f54a45';
exports.caltoken_event_light_bg_bar_tur = '#10a893';
exports.caltoken_event_light_bg_bar_violet = '#cf5ecf';
exports.caltoken_event_light_bg_bar_wathet = '#1295ca';
exports.caltoken_event_light_bg_bar_yellow = '#d99904';
exports.caltoken_event_light_bg_blue = '#e0e9ff';
exports.caltoken_event_light_bg_carmine = '#fee2f2';
exports.caltoken_event_light_bg_green = '#d0f5ce';
exports.caltoken_event_light_bg_indigo = '#e9eafb';
exports.caltoken_event_light_bg_neutral = '#dee0e3';
exports.caltoken_event_light_bg_oran = '#fee7cd';
exports.caltoken_event_light_bg_pending_blue = '#f0f4ff';
exports.caltoken_event_light_bg_pending_carmine = '#fef0f8';
exports.caltoken_event_light_bg_pending_green = '#e4fae1';
exports.caltoken_event_light_bg_pending_indigo = '#f2f3fd';
exports.caltoken_event_light_bg_pending_neutral = '#eff0f1';
exports.caltoken_event_light_bg_pending_orange = '#fff3e5';
exports.caltoken_event_light_bg_pending_purple = '#f5f0fe';
exports.caltoken_event_light_bg_pending_red = '#fef0f0';
exports.caltoken_event_light_bg_pending_tur = '#e2f8f5';
exports.caltoken_event_light_bg_pending_violet = '#fceefc';
exports.caltoken_event_light_bg_pending_wathet = '#e7f8fe';
exports.caltoken_event_light_bg_pending_yellow = '#fbf4df';
exports.caltoken_event_light_bg_purple = '#efe6fe';
exports.caltoken_event_light_bg_red = '#fee3e2';
exports.caltoken_event_light_bg_tur = '#c4f2ec';
exports.caltoken_event_light_bg_violet = '#f9e2f9';
exports.caltoken_event_light_bg_wathet = '#caeffc';
exports.caltoken_event_light_bg_yellow = '#faedc2';
exports.caltoken_event_light_calendar_blue = '#4e83fd';
exports.caltoken_event_light_calendar_carmine = '#f14ba9';
exports.caltoken_event_light_calendar_green = '#34c724';
exports.caltoken_event_light_calendar_indigo = '#616ae5';
exports.caltoken_event_light_calendar_neutral = '#9ca2a9';
exports.caltoken_event_light_calendar_orange = '#ff8800';
exports.caltoken_event_light_calendar_purple = '#935af6';
exports.caltoken_event_light_calendar_red = '#f76964';
exports.caltoken_event_light_calendar_tur = '#00d6b9';
exports.caltoken_event_light_calendar_violet = '#d136d1';
exports.caltoken_event_light_calendar_wathet = '#14c0ff';
exports.caltoken_event_light_calendar_yellow = '#ffc60a';
exports.caltoken_event_light_text_blue = '#0442d2';
exports.caltoken_event_light_text_carmine = '#9d1562';
exports.caltoken_event_light_text_green = '#0b6017';
exports.caltoken_event_light_text_indigo = '#333dcc';
exports.caltoken_event_light_text_neutral = '#51565d';
exports.caltoken_event_light_text_orange = '#a44904';
exports.caltoken_event_light_text_purple = '#4811a6';
exports.caltoken_event_light_text_red = '#a11c17';
exports.caltoken_event_light_text_tur = '#045d51';
exports.caltoken_event_light_text_unanswered_blue = '#3668db';
exports.caltoken_event_light_text_unanswered_carmine = '#b14481';
exports.caltoken_event_light_text_unanswered_green = '#388c2e';
exports.caltoken_event_light_text_unanswered_indigo = '#525ad4';
exports.caltoken_event_light_text_unanswered_neutral = '#646a73';
exports.caltoken_event_light_text_unanswered_orange = '#b66d36';
exports.caltoken_event_light_text_unanswered_purple = '#6d41b8';
exports.caltoken_event_light_text_unanswered_red = '#b85451';
exports.caltoken_event_light_text_unanswered_tur = '#067062';
exports.caltoken_event_light_text_unanswered_violet = '#9f529f';
exports.caltoken_event_light_text_unanswered_wathet = '#047fb0';
exports.caltoken_event_light_text_unanswered_yellow = '#ad7a03';
exports.caltoken_event_light_text_violet = '#872787';
exports.caltoken_event_light_text_wathet = '#076a94';
exports.caltoken_event_light_text_yellow = '#865b03';
exports.caltoken_event_quick_creation_shadow = '0px 10px 36px 10px rgba(31, 35, 41, 0.09), 0px 8px 24px rgba(31, 35, 41, 0.09), 0px 6px 12px -10px rgba(31, 35, 41, 0.11)';
exports.caltoken_event_stripe_blue = '#f0f4ff';
exports.caltoken_event_stripe_carmine = '#fef0f8';
exports.caltoken_event_stripe_green = '#e4fae1';
exports.caltoken_event_stripe_indigo = '#f2f3fd';
exports.caltoken_event_stripe_neutral = '#f2f3f5';
exports.caltoken_event_stripe_orange = '#fff3e5';
exports.caltoken_event_stripe_purple = '#f5f0fe';
exports.caltoken_event_stripe_red = '#fef0f0';
exports.caltoken_event_stripe_tur = '#e2f8f5';
exports.caltoken_event_stripe_violet = '#fceefc';
exports.caltoken_event_stripe_wathet = '#e7f8fe';
exports.caltoken_event_stripe_yellow = '#fbf4df';
exports.caltoken_tag_colour_day = '#e29b00';
