// @byted-service/logagent/lib/flusher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Flusher = void 0;
class Flusher {
    constructor() {
        this.flushableMap = new Map();
        this.intervals = new Map();
    }
    register(key, flushable) {
        const flushables = this.flushableMap.get(key);
        if (flushables) {
            flushables.add(flushable);
        }
        else {
            this.flushableMap.set(key, new Set([flushable]));
            this.intervals.set(key, setInterval(this.handle.bind(this), key, key).unref());
        }
    }
    close() {
        for (const [, intervalId] of this.intervals) {
            clearInterval(intervalId);
        }
        for (const [, flushables] of this.flushableMap) {
            for (const flushable of flushables) {
                try {
                    flushable.flush();
                }
                catch (err) {
                    // do nothing
                }
            }
        }
        this.flushableMap = new Map();
        this.intervals = new Map();
    }
    handle(key) {
        const flushables = this.flushableMap.get(key);
        if (!flushables || flushables.size === 0) {
            const intervalId = this.intervals.get(key);
            if (intervalId) {
                clearInterval(intervalId);
            }
            this.flushableMap.delete(key);
            this.intervals.delete(key);
            return;
        }
        const failedFlushables = new Set();
        for (const flushable of flushables) {
            try {
                flushable.flush();
            }
            catch (err) {
                failedFlushables.add(flushable);
            }
        }
        this.flushableMap.set(key, failedFlushables);
    }
}
exports.Flusher = Flusher;
