// @byted-service/logagent/lib/logagent_v2.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LogagentV2 = void 0;
const tslib_1 = require("tslib");
const env_1 = require("@byted-service/env");
const metrics_1 = tslib_1.__importDefault(require("@byted-service/metrics"));
const flusher_1 = require("./flusher");
const protocol_1 = require("./protocol");
const metrics_v2_1 = require("./metrics_v2");
const sender_1 = require("./sender");
const helper_1 = require("./helper");
const consts_1 = require("./consts");
const flusher = new flusher_1.Flusher();
/**
 * 目前还在试验阶段，外部请勿使用
 */
class LogagentV2 {
    constructor(options = {}) {
        this.length = 0;
        this.msgs = [];
        const { psm, flushInterval, maxLength, host, port, socketPath, logger, metrics, } = options;
        this.psm = psm || (0, env_1.getPSM)();
        this.flushInterval = flushInterval || consts_1.defaultFlushInterval;
        this.maxLength = maxLength || consts_1.maxLen;
        this.logger = logger || consts_1.defaultLogger;
        this.metrics = (0, metrics_v2_1.wrapMetrics)(metrics ||
            new metrics_1.default({
                defaultPrefix: 'toutiao.infra.ttlogagent.nodejssdk_v1',
                logger: this.logger,
            }), this);
        this.sendMsg = (0, sender_1.getMessageSender)({
            host,
            port,
            socketPath,
            logger: this.logger,
        });
    }
    send(msg, tags = {}, tagKeys) {
        const data = (0, helper_1.truncate)(typeof msg === 'string' ? Buffer.from(msg) : msg, consts_1.maxPacketMsgLen);
        /**
         * https://docs.bytedance.net/doc/uhnP66an8j3qU0mn2Kjywd
         * https://code.byted.org/log_market/ttlogagent/blob/master/agent/formatter/text_formatter.go#L110
         *
         * 数据格式：model.Msg
         * 最终结果：_level _ts _location _host _psm _logid _cluster _deployStage _idc _spanID ...tags _msg=日志内容
         */
        tags._taskName = tags._taskName || this.psm;
        tags._language = tags._language || 'js';
        tags._process = tags._process || process.pid;
        tags._level = tags._level || 'INFO';
        tags._ts = tags._ts || Date.now();
        tags._location = tags._location || 'byted-service/logagent.js:0';
        tags._logid = tags._logid || '-';
        tags._spanID = tags._spanID || 0;
        tags._psm = tags._psm || this.psm;
        tags._cluster = tags._cluster || (0, env_1.getCluster)();
        tags._idc = tags._idc || (0, env_1.getIDC)();
        tags._deployStage = tags._deployStage || (0, env_1.getStage)();
        tags._host = tags._host || (0, env_1.getIP)();
        tags._podName = tags._podName || (0, env_1.getPodName)();
        tagKeys = tagKeys || Object.keys(tags);
        const kvs = [];
        const mark = {};
        for (let i = 0; i < tagKeys.length; i++) {
            const key = tagKeys[i];
            const value = tags[key];
            if (value === undefined) {
                continue;
            }
            kvs.push({ key, value: String(value) });
            mark[key] = true;
        }
        for (let i = 0; i < consts_1.builtinTagKeys.length; i++) {
            const key = consts_1.builtinTagKeys[i];
            const value = tags[key];
            if (!mark[key]) {
                kvs.push({ key, value: String(value) });
            }
        }
        let error = null;
        try {
            const msg = (0, protocol_1.quickEncode)(data, kvs);
            this.length += msg.byteLength;
            this.msgs.push(msg);
            flusher.register(this.flushInterval, this);
            if (this.length >= this.maxLength) {
                this.flush();
            }
            return;
        }
        catch (err) {
            error = err;
        }
        if (!error.message.includes('packet is too long')) {
            this.metrics.encodeError();
            throw error;
        }
        const byteLength = data.byteLength - error[consts_1.byteLengthSym] + consts_1.maxPacketLen;
        if (byteLength < 0) {
            this.metrics.packetTooLongError();
            throw error;
        }
        try {
            const msg = (0, protocol_1.quickEncode)((0, helper_1.truncate)(data, byteLength), kvs);
            this.length += msg.byteLength;
            this.msgs.push(msg);
            flusher.register(this.flushInterval, this);
            if (this.length >= this.maxLength) {
                this.flush();
            }
            return;
        }
        catch (err) {
            error = err;
        }
        this.metrics.encodeError();
        throw error;
    }
    flush() {
        if (this.length === 0) {
            return;
        }
        const msgs = this.msgs;
        this.length = 0;
        this.msgs = [];
        this.sendMsg(msgs).catch(consts_1.noop);
    }
}
exports.LogagentV2 = LogagentV2;
