// @byted-service/logagent/lib/sender.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMessageSender = exports.sendViaUnixDomainSocket = exports.sendViaHTTPProtocol = void 0;
const tslib_1 = require("tslib");
const net_1 = require("net");
const fs_1 = require("fs");
const node_fetch_1 = tslib_1.__importDefault(require("node-fetch"));
const pool_1 = require("./pool");
const helper_1 = require("./helper");
const consts_1 = require("./consts");
const sendViaHTTPProtocol = (host, port) => async function (msgs) {
    // 由于这里的 json 格式是固定的，所以直接使用字符串拼接，以提升性能
    const payload = `{"data":[${msgs
        .map((msg) => `[${msg.join(',')}]`)
        .join(',')}]}`;
    try {
        await (0, node_fetch_1.default)(`http://${host}:${port}`, {
            method: 'PUT',
            body: payload,
            headers: {
                'content-type': 'application/json',
            },
        });
        /**
         * 整体代码重构会有大量改动，所以这里先暂时用 this 来解决
         */
        this.metrics.success(msgs);
    }
    catch (err) {
        /**
         * 整体代码重构会有大量改动，所以这里先暂时用 this 来解决
         */
        this.metrics.sendHttpError(msgs);
    }
};
exports.sendViaHTTPProtocol = sendViaHTTPProtocol;
const sendViaUnixDomainSocket = (socketPath, logger) => async function (msgs) {
    const successMessages = [];
    const failedMessages = [];
    // TODO pipe socket失效的期间可以考虑改用http发送日志？
    const sock = await pool_1.pool.acquire({ socketPath, logger });
    for (let i = 0; i < msgs.length; i++) {
        const msg = msgs[i];
        const header = Buffer.alloc(8);
        header.writeInt32BE(0x10100000, 0);
        header.writeInt32BE(msg.byteLength, 4);
        try {
            sock.write(header);
            sock.write(
            // 兼容 node 8
            Buffer.from(msg));
            successMessages.push(msgs[i]);
        }
        catch (err) {
            /* istanbul ignore next */
            failedMessages.push(msgs[i]);
        }
    }
    pool_1.pool.release(sock);
    if (successMessages.length > 0) {
        /**
         * 整体代码重构会有大量改动，所以这里先暂时用 this 来解决
         */
        this.metrics.success(successMessages);
    }
    /* istanbul ignore next */
    if (failedMessages.length > 0) {
        /**
         * 整体代码重构会有大量改动，所以这里先暂时用 this 来解决
         */
        this.metrics.sendSocketError(failedMessages);
    }
};
exports.sendViaUnixDomainSocket = sendViaUnixDomainSocket;
function getMessageSender(options) {
    const gateway = (0, helper_1.getGateway)();
    let socketPath = '';
    let host = '';
    let port = 0;
    switch (true) {
        case typeof options.socketPath === 'string':
            socketPath = options.socketPath;
            return (0, exports.sendViaUnixDomainSocket)(socketPath, options.logger);
        case typeof options.host === 'string' &&
            typeof options.port === 'number':
            host = options.host;
            host = (0, net_1.isIPv6)(host) ? `[${host}]` : host;
            port = options.port;
            return (0, exports.sendViaHTTPProtocol)(host, port);
        case (0, fs_1.existsSync)(consts_1.ttlogagentSocketPath):
            socketPath = consts_1.ttlogagentSocketPath;
            return (0, exports.sendViaUnixDomainSocket)(socketPath, options.logger);
        default:
            host = (0, net_1.isIPv6)(gateway) ? `[${gateway}]` : gateway;
            port = consts_1.ttlogagentHTTPPort;
            return (0, exports.sendViaHTTPProtocol)(host, port);
    }
}
exports.getMessageSender = getMessageSender;
