// @universe-design/styles/ts/ccm-ccm-chat/index.dark.atom-raw.ts
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
export const ccmtoken_chart_aurora_b100_raw = '189, 210, 255';
export const ccmtoken_chart_aurora_b200_raw = '143, 180, 255';
export const ccmtoken_chart_aurora_b300_raw = '76, 136, 255';
export const ccmtoken_chart_aurora_b400_raw = '51, 112, 235';
export const ccmtoken_chart_aurora_b500_raw = '38, 85, 182';
export const ccmtoken_chart_aurora_b600_raw = '25, 66, 148';
export const ccmtoken_chart_aurora_c100_raw = '255, 194, 229';
export const ccmtoken_chart_aurora_c200_raw = '252, 148, 207';
export const ccmtoken_chart_aurora_c300_raw = '237, 119, 186';
export const ccmtoken_chart_aurora_c400_raw = '194, 74, 142';
export const ccmtoken_chart_aurora_c500_raw = '148, 56, 108';
export const ccmtoken_chart_aurora_c600_raw = '120, 43, 87';
export const ccmtoken_chart_aurora_g100_raw = '153, 228, 144';
export const ccmtoken_chart_aurora_g200_raw = '105, 204, 92';
export const ccmtoken_chart_aurora_g300_raw = '81, 186, 67';
export const ccmtoken_chart_aurora_g400_raw = '65, 158, 52';
export const ccmtoken_chart_aurora_g500_raw = '47, 117, 38';
export const ccmtoken_chart_aurora_g600_raw = '33, 81, 26';
export const ccmtoken_chart_aurora_o100_raw = '254, 200, 139';
export const ccmtoken_chart_aurora_o200_raw = '248, 158, 68';
export const ccmtoken_chart_aurora_o300_raw = '243, 135, 27';
export const ccmtoken_chart_aurora_o400_raw = '219, 112, 24';
export const ccmtoken_chart_aurora_o500_raw = '161, 83, 23';
export const ccmtoken_chart_aurora_o600_raw = '104, 58, 18';
export const ccmtoken_chart_aurora_p100_raw = '219, 200, 253';
export const ccmtoken_chart_aurora_p200_raw = '197, 163, 255';
export const ccmtoken_chart_aurora_p300_raw = '165, 117, 250';
export const ccmtoken_chart_aurora_p400_raw = '140, 85, 236';
export const ccmtoken_chart_aurora_p500_raw = '108, 57, 198';
export const ccmtoken_chart_aurora_p600_raw = '85, 41, 163';
export const ccmtoken_chart_aurora_r100_raw = '248, 158, 155';
export const ccmtoken_chart_aurora_r200_raw = '246, 130, 126';
export const ccmtoken_chart_aurora_r300_raw = '240, 91, 86';
export const ccmtoken_chart_aurora_r400_raw = '209, 70, 66';
export const ccmtoken_chart_aurora_r500_raw = '160, 51, 49';
export const ccmtoken_chart_aurora_r600_raw = '160, 51, 49';
export const ccmtoken_chart_aurora_t100_raw = '111, 232, 216';
export const ccmtoken_chart_aurora_t200_raw = '51, 214, 192';
export const ccmtoken_chart_aurora_t300_raw = '45, 190, 171';
export const ccmtoken_chart_aurora_t400_raw = '16, 168, 147';
export const ccmtoken_chart_aurora_t500_raw = '15, 133, 117';
export const ccmtoken_chart_aurora_t600_raw = '29, 78, 71';
export const ccmtoken_chart_aurora_w100_raw = '137, 223, 254';
export const ccmtoken_chart_aurora_w200_raw = '81, 195, 240';
export const ccmtoken_chart_aurora_w300_raw = '37, 178, 229';
export const ccmtoken_chart_aurora_w400_raw = '16, 153, 204';
export const ccmtoken_chart_aurora_w500_raw = '20, 108, 145';
export const ccmtoken_chart_aurora_w600_raw = '22, 67, 89';
export const ccmtoken_chart_aurora_y100_raw = '255, 222, 117';
export const ccmtoken_chart_aurora_y200_raw = '251, 203, 70';
export const ccmtoken_chart_aurora_y300_raw = '240, 182, 34';
export const ccmtoken_chart_aurora_y400_raw = '217, 153, 4';
export const ccmtoken_chart_aurora_y500_raw = '173, 122, 3';
export const ccmtoken_chart_aurora_y600_raw = '111, 74, 1';
