// @universe-design/react/src/details-display/DetailsDisplayContext.ts
import type * as React from 'react';
import { createContextHook } from '@universe-design/react-utilities';
import type { DefaultSizeType } from '../util';
import type { DetailsDisplayDataItem } from '../details-display';
import { detailsDisplayDefaults } from './defaults';
import { useCssScope } from '../config-provider';

export interface UseDetailsDisplayProps {
  /**
   * 每行展示的列数
   * @defaultValue 3
   */
  column?: number;
  /**
   * 详情展示组件的布局方式
   * @defaultValue 'horizontal'
   */
  layout?: 'horizontal' | 'vertical' | 'inline';
  /**
   * 列表大小
   * @defaultValue 'medium'
   */
  size?: DefaultSizeType;
  /**
   * 是否显示冒号
   * @defaultValue true
   */
  colon?: boolean;
  /**
   * 自定义标签样式
   * @defaultValue `{}`
   */
  labelStyle?: React.CSSProperties;
  /**
   * 自定义内容样式
   * @defaultValue `{}`
   */
  contentStyle?: React.CSSProperties;
  /**
   * 自定义标签类名
   * @defaultValue ''
   */
  labelClassName?: string;
  /**
   * 自定义内容类名
   * @defaultValue ''
   */
  contentClassName?: string;
  /**
   * 详情展示组件的对齐方式
   * @defaultValue 'start'
   */
  align?: 'center' | 'justify' | 'start';
  /**
   * 列表显示的数据
   * @defaultValue `[]`
   */
  data?: DetailsDisplayDataItem[];
}

/**
 * detailsDisplay hook
 */
export const useDetailsDisplay = (props: UseDetailsDisplayProps) => {
  const {
    column = detailsDisplayDefaults.column,
    layout = detailsDisplayDefaults.layout,
    size = detailsDisplayDefaults.size,
    colon = detailsDisplayDefaults.colon,
    align = detailsDisplayDefaults.align,
    data = detailsDisplayDefaults.data,
    labelStyle = detailsDisplayDefaults.labelStyle,
    contentStyle = detailsDisplayDefaults.contentStyle,
    labelClassName = detailsDisplayDefaults.labelClassName,
    contentClassName = detailsDisplayDefaults.contentClassName,
  } = props;

  // prefix
  const prefix = useCssScope('details-display')();

  // align
  const mergedAlign = (() => {
    if (layout === 'vertical' || layout === 'inline') {
      return 'start';
    }
    return align;
  })();

  return {
    prefix,
    column,
    layout,
    size,
    colon,
    align: mergedAlign,
    data,
    labelStyle,
    contentStyle,
    labelClassName,
    contentClassName,
  };
};

export type DetailsDisplayContextType = ReturnType<typeof useDetailsDisplay>;

/**
 * detailsDisplay 组件全局状态
 */
export const { Provider: DetailsDisplayContextProvider, useContext: useDetailsDisplayContext } =
  createContextHook<DetailsDisplayContextType>('DetailsDisplayContext', {
    strict: true,
  });
