// @universe-design/react/src/details-display/DetailsDisplay.tsx
/** @jsxRuntime classic */
/** @jsx jsx */
import * as React from 'react';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx } from '@universe-design/stylex/react';
import type { StylexProps } from '../util';
import type { UseDetailsDisplayProps } from './DetailsDisplayContext';
import { DetailsDisplayContextProvider, useDetailsDisplay } from './DetailsDisplayContext';
import { useDetailsDisplayStylexStyles } from './sx';
import Row from './Row';

/**
 * @docgen
 */
export interface DetailsDisplayProps extends UseDetailsDisplayProps, StylexProps {
  /**
   * 详情展示组件的样式
   */
  style?: React.CSSProperties;
  /**
   * 详情展示组件的类名
   */
  className?: string;
}

/**
 * @docgen
 */
export interface DetailsDisplayDataItem {
  /**
   * 内容的描述
   * @defaultValue '--'
   */
  label?: React.ReactNode;
  /**
   * 内容
   * @defaultValue '--'
   */
  content?: (() => React.ReactNode) | React.ReactNode;
  /**
   * 自定义标签样式
   * @defaultValue `{}`
   */
  labelStyle?: React.CSSProperties;
  /**
   * 自定义内容样式
   *  @defaultValue `{}`
   */
  contentStyle?: React.CSSProperties;
  /**
   * 自定义标签类名
   *  @defaultValue ''
   */
  labelClassName?: string;
  /**
   * 自定义内容类名
   *  @defaultValue ''
   */
  contentClassName?: string;
  /**
   * 包含列的数量
   *  @defaultValue 1
   */
  span?: number;
}

export const DetailsDisplay = (props: DetailsDisplayProps) => {
  const { style, className, sx = [], ...useDetailsDisplayProp } = props;
  const contextValue = useDetailsDisplay(useDetailsDisplayProp);
  const { prefix, column, data } = contextValue;

  /**
   * format
   * 1、把每个数据的 span 格式化为 1 或者最大值 column；
   * 2、二维数组的每一个一维数组为一个tr；
   * 3、每一个数据的 span 大于 column 或者在每行的剩余空间放不下时，自动放在下一行；
   * 3.1、放到下一行后，会把上一行的最后一个数据扩大为占满剩余 tr 空间 span；
   * 4、最后输出二维数组；
   */
  const getRowItems = (originData: DetailsDisplayDataItem[], _column: number) => {
    const rows: DetailsDisplayDataItem[][] = [];
    let tmpRow: DetailsDisplayDataItem[] = [];
    let rowRestCol: number = _column;

    originData.forEach((item, index) => {
      const span: number = Math.min(item.span || 1, _column);
      item.span = span;

      if (span <= rowRestCol) {
        tmpRow.push(item);
        rowRestCol -= span;
      } else {
        rows.push(tmpRow);
        tmpRow = [item];
        rowRestCol = _column - span;
      }

      if (rowRestCol === 0 || index === originData.length - 1) {
        rows.push(tmpRow);
        tmpRow = [];
        rowRestCol = _column;
      }
    });

    rows.forEach((item: DetailsDisplayDataItem[]) => {
      const sum: number = item.reduce((acc: number, child: DetailsDisplayDataItem) => acc + (child.span || 1), 0);
      if (sum < _column) {
        item[item.length - 1].span! += _column - sum;
      }
    });

    return rows;
  };

  const rowItems = getRowItems(data, column);

  const styles = useDetailsDisplayStylexStyles({ prefix });

  return (
    <DetailsDisplayContextProvider value={contextValue}>
      <div className={className} sx={[styles.base, ...sx]} style={style}>
        <table sx={[styles.table]}>
          <tbody>
            {rowItems.map((row, r) => (
              <Row key={r} isRowEnd={r === rowItems.length - 1} row={row} />
            ))}
          </tbody>
        </table>
      </div>
    </DetailsDisplayContextProvider>
  );
};
