// @universe-design/react/src/details-display/DetailsDisplayItemContext.ts
import { createContextHook } from '@universe-design/react-utilities';
import type { DetailsDisplayDataItem } from '../details-display';
import { detailsDisplayItemDefaults } from './defaults';

export interface UseDetailsDisplayItemProps extends DetailsDisplayDataItem {
  /**
   * @internal
   * 是否是最后一行
   */
  isRowEnd: boolean;
  /**
   * @internal
   * 是否是最后一列
   */
  isColumnEnd?: boolean;
}

/**
 * detailsDisplayItem hook
 */
export const useDetailsDisplayItem = (props: UseDetailsDisplayItemProps) => {
  const {
    label = detailsDisplayItemDefaults.label,
    content = detailsDisplayItemDefaults.content,
    labelStyle = detailsDisplayItemDefaults.labelStyle,
    contentStyle = detailsDisplayItemDefaults.contentStyle,
    labelClassName = detailsDisplayItemDefaults.labelClassName,
    contentClassName = detailsDisplayItemDefaults.contentClassName,
    span = detailsDisplayItemDefaults.span,
    isRowEnd = detailsDisplayItemDefaults.isRowEnd,
    isColumnEnd = detailsDisplayItemDefaults.isColumnEnd,
  } = props;

  return {
    label,
    content,
    labelStyle,
    contentStyle,
    labelClassName,
    contentClassName,
    span,
    isRowEnd,
    isColumnEnd,
  };
};

export type DetailsDisplayItemContextType = ReturnType<typeof useDetailsDisplayItem>;

/**
 * detailsDisplayItem 组件状态
 */
export const { Provider: DetailsDisplayItemContextProvider, useContext: useDetailsDisplayItemContext } =
  createContextHook<DetailsDisplayItemContextType>('DetailsDisplayItemContext', {
    strict: true,
  });
