// @universe-design/react/src/details-display/Row.tsx
/* eslint-disable react/no-unknown-property */
/** @jsxRuntime classic */
/** @jsx jsx */
/** @jsxFrag React.Fragment */
import * as React from 'react';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx } from '@universe-design/stylex/react';
import type { DetailsDisplayDataItem } from './DetailsDisplay';
import { useDetailsDisplayContext } from './DetailsDisplayContext';
import { DetailsDisplayItemContextProvider, useDetailsDisplayItem } from './DetailsDisplayItemContext';
import { useRowStylexStyles } from './sx';
import HeaderCell from './HeaderCell';
import DataCell from './DataCell';
import Label from './Label';
import Content from './Content';

export interface CellConfig {
  /**
   * @internal
   * 类型
   */
  type: string;
  /**
   * @internal
   * 渲染的标签
   */
  component: string | [string, string];
  /**
   * @internal
   * 是否显示描述
   */
  showLabel?: boolean;
  /**
   * @internal
   * 是否显示内容
   */
  showContent?: boolean;
  /**
   * @internal
   * 是否是最后一行
   */
  isRowEnd: boolean;
}

const renderCells = (
  items: DetailsDisplayDataItem[],
  { type, component, showLabel, showContent, isRowEnd }: CellConfig,
) =>
  items.map((item: DetailsDisplayDataItem, i: number) => {
    const { layout } = useDetailsDisplayContext();
    const layoutInline = layout === 'inline';
    // 处理cell之间的间隙，是否是最后一列
    const isColumnEnd = i === items.length - 1;
    const contextValue = useDetailsDisplayItem({ ...item, isRowEnd, isColumnEnd });

    if (typeof component === 'string') {
      return (
        <DetailsDisplayItemContextProvider key={`${type}-${component}-${i}`} value={contextValue}>
          {showLabel && !layoutInline && (
            <HeaderCell>
              <Label />
            </HeaderCell>
          )}
          {showContent && (
            <DataCell>
              {layoutInline && <Label />}
              <Content />
            </DataCell>
          )}
        </DetailsDisplayItemContextProvider>
      );
    }

    return [
      <DetailsDisplayItemContextProvider key={`${type}-${component[0]}-${i}`} value={{ ...contextValue, span: 1 }}>
        <HeaderCell>
          <Label />
        </HeaderCell>
      </DetailsDisplayItemContextProvider>,
      <DetailsDisplayItemContextProvider
        key={`${type}-${component[1]}-${i}`}
        value={{ ...contextValue, span: (item.span || 1) * 2 - 1 }}
      >
        <DataCell>
          {layoutInline && <Label />}
          <Content />
        </DataCell>
      </DetailsDisplayItemContextProvider>,
    ];
  });

export interface RowProps {
  /**
   * @internal
   * 是否是最后一行
   */
  isRowEnd: boolean;
  /**
   * @internal
   * 每行的数据
   */
  row: DetailsDisplayDataItem[];
}

const Row = (props: RowProps) => {
  const { isRowEnd, row } = props;
  const { prefix, layout } = useDetailsDisplayContext();

  const styles = useRowStylexStyles({ prefix: `${prefix}__row` });

  if (layout === 'vertical') {
    return (
      <>
        <tr sx={[styles.base]}>
          {renderCells(row, {
            type: 'label',
            component: 'th',
            showLabel: true,
            isRowEnd: false,
          })}
        </tr>
        <tr sx={[styles.base]}>
          {renderCells(row, {
            type: 'content',
            component: 'td',
            showContent: true,
            isRowEnd,
          })}
        </tr>
      </>
    );
  }

  if (layout === 'inline') {
    return (
      <tr sx={[styles.base]}>
        {renderCells(row, {
          type: 'inline',
          component: 'td',
          showLabel: true,
          showContent: true,
          isRowEnd,
        })}
      </tr>
    );
  }

  return (
    <tr sx={[styles.base]}>
      {renderCells(row, {
        type: 'item',
        component: ['th', 'td'],
        isRowEnd,
      })}
    </tr>
  );
};

export default Row;
