// @universe-design/react/src/details-display/sx.ts
import stylex from '@universe-design/stylex';
import { makeStylexStyles } from '@universe-design/stylex/react';

// table
export const useDetailsDisplayStylexStyles = makeStylexStyles(({ prefix }: { prefix: string }) => ({
  base: stylex.create(
    {
      fontSize: 14,
    },
    prefix,
  ),
  table: stylex.create(
    {
      width: '100%',
    },
    `${prefix}__table`,
  ),
}));

// row
export const useRowStylexStyles = makeStylexStyles(({ prefix }: { prefix: string }) => ({
  base: stylex.create(
    {
      lineHeight: '22px',
      verticalAlign: 'baseline',
    },
    prefix,
  ),
}));

// HeaderCell && DataCell
export const useCellStylexStyles = makeStylexStyles(({ prefix }: { prefix: string }) => ({
  /**
   * th & reset
   * minWidth：两个汉字（14 * 2） + 冒号的宽度（16）
   *  */
  th: stylex.create(
    {
      padding: 0,
      minWidth: 44,
    },
    `${prefix}`,
  ),
  // td & reset
  td: stylex.create(
    {
      padding: 0,
      color: $token['text-title'],
      minWidth: 44,
    },
    `${prefix}`,
  ),
  // 列表三种尺寸间距
  size: {
    small: stylex.create(
      {
        paddingBottom: 12,
      },
      `${prefix}-small`,
    ),
    medium: stylex.create(
      {
        paddingBottom: 16,
      },
      `${prefix}-medium`,
    ),
    large: stylex.create(
      {
        paddingBottom: 20,
      },
      `${prefix}-large`,
    ),
  },
  align: {
    left: stylex.create(
      {
        textAlign: 'left',
      },
      `${prefix}-align-left`,
    ),
    right: stylex.create(
      {
        textAlign: 'right',
      },
      `${prefix}-align-right`,
    ),
  },
  // 垂直布局 label 和 content 的间距
  headerGap: stylex.create(
    {
      paddingBottom: 4,
    },
    `${prefix}-gap`,
  ),
  layout: {
    // 垂直布局横向 item 间距
    verticalGap: stylex.create(
      {
        paddingRight: 24,
      },
      `${prefix}-vertical-gap`,
    ),
    // 水平布局横向 item 间距
    horizontalGap: stylex.create(
      {
        paddingRight: 40,
      },
      `${prefix}-horizontal-gap`,
    ),
  },
  // layout === 'inline'时，设置display: 'inline-flex'
  inlineFlex: stylex.create(
    {
      display: 'inline-flex',
      alignItems: 'baseline',
    },
    `${prefix}-inline-flex`,
  ),
}));

// label
export const useLabelStylexStyles = makeStylexStyles(({ prefix }: { prefix: string }) => ({
  base: stylex.create(
    {
      display: 'inline-flex',
      alignItems: 'baseline',
      color: $token['text-caption'],
    },
    prefix,
  ),
  colon: stylex.create(
    {
      width: 16,
      minWidth: 16,
      textAlign: 'left',
    },
    `${prefix}__colon`,
  ),
}));
