// @universe-design/react/lib/transfer/Transfer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Transfer = void 0;
var React = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _List = require("../shared/components/list/List");
var _TransferContext = require("./hooks/TransferContext");
var _configProvider = require("../config-provider");
var _SourcePanel = require("./panel/SourcePanel");
var _TargetPanel = require("./panel/TargetPanel");
var _sx = require("./sx");
const Transfer = exports.Transfer = /*#__PURE__*/React.forwardRef((props, ref) => {
  const cssScope = (0, _configProvider.useCssScope)('transfer');
  const defaultChildren = React.useMemo(() => /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(_SourcePanel.TransferSourcePanel, null), /*#__PURE__*/React.createElement(_TargetPanel.TransferTargetPanel, null)), []);
  const {
    children = defaultChildren,
    style,
    className,
    ...useTransferProps
  } = props;
  const ctxValue = (0, _TransferContext.useTransfer)(useTransferProps);
  React.useImperativeHandle(ref, () => ({
    clearSearchValue: () => ctxValue.setSearchValue('')
  }));
  return /*#__PURE__*/React.createElement(_TransferContext.TransferProvider, {
    value: ctxValue
  }, /*#__PURE__*/React.createElement(_List.ListBox, {
    cascaded: true,
    style: style,
    sx: [_sx.sxTransfer],
    className: (0, _clsx.default)(cssScope(), className)
  }, children));
});
