// @universe-design/react/src/context-menu/stories/homeless.stories.tsx
import { css } from '@emotion/css';
import React from 'react';
import { contextMenu } from '..';
import { Menu, Dropdown, ConfigProvider } from '../..';
import { FloatingUIPopperSidecar } from '../../shared';

export default {
  title: 'Components/ContextMenu/临时示例',
};

const link = 'https://universe.bytedance.net/';

const contextAreaCss = css({
  display: 'flex',
  flex: 1,
  height: 300,
  backgroundColor: 'rgba(0, 0, 0, .2)',
  alignItems: 'center',
  justifyContent: 'center',
});

export const Demo = () => {
  const menu = (
    <Menu>
      <Menu.Item>
        <a target="_blank" rel="noopener noreferrer" href={link}>
          1st menu item
        </a>
      </Menu.Item>
      <Menu.Item>
        <a target="_blank" rel="noopener noreferrer" href={link}>
          2nd menu item
        </a>
      </Menu.Item>
      <Menu.Item>
        <a target="_blank" rel="noopener noreferrer" href={link}>
          3rd menu item
        </a>
      </Menu.Item>
    </Menu>
  );

  const open = (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
    e.preventDefault();
    console.log('ContextMenu', [e.pageX, e.pageY]);
    contextMenu.open({
      menu,
      position: [e.pageX, e.pageY],
      onClose: () => console.log('onClose'),
    });
  };

  return (
    <div
      style={{
        display: 'flex',
        gap: 16,
      }}
    >
      <div className={contextAreaCss} onContextMenu={open}>
        点
      </div>
      <ConfigProvider>
        <Dropdown trigger="contextMenu" overlay={menu}>
          <div className={contextAreaCss}>点（RcTrigger）</div>
        </Dropdown>
      </ConfigProvider>
      <ConfigProvider PopperSidecar={FloatingUIPopperSidecar}>
        <Dropdown trigger="contextMenu" overlay={menu}>
          <div className={contextAreaCss}>点（Popper）</div>
        </Dropdown>
      </ConfigProvider>
    </div>
  );
};
