// @universe-design/styles/esm/base/index.light.atom.js
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
export const basetoken_AI_instruction_hover = 'rgba(31, 35, 41, 0.06)';
export const basetoken_AI_instruction_normal = '#ffffff';
export const basetoken_AI_instruction_press = 'rgba(31, 35, 41, 0.1)';
export const basetoken_container_bg = '#f5f6f7';
export const basetoken_dashboard_sidebar_bg_overlay = '#f8f9fa';
export const basetoken_dashboard_slicer_tag_bg_hover = 'rgba(255, 255, 255, 0.5)';
export const basetoken_dashboard_slicer_tag_bg_normal = 'rgba(255, 255, 255, 0.9)';
export const basetoken_form_award = '#465399';
export const basetoken_form_award_01 = 'rgba(70, 83, 153, 0.64)';
export const basetoken_form_award_02 = 'rgba(70, 83, 153, 0.4)';
export const basetoken_form_award_03 = 'rgba(70, 83, 153, 0.24)';
export const basetoken_form_award_04 = 'rgba(70, 83, 153, 0.08)';
export const basetoken_form_desert = '#6a624e';
export const basetoken_form_desert_01 = 'rgba(106, 98, 78, 0.64)';
export const basetoken_form_desert_02 = 'rgba(106, 98, 78, 0.4)';
export const basetoken_form_desert_03 = 'rgba(106, 98, 78, 0.24)';
export const basetoken_form_desert_04 = 'rgba(106, 98, 78, 0.08)';
export const basetoken_form_entertainment = '#5d6545';
export const basetoken_form_entertainment_01 = 'rgba(93, 101, 69, 0.64)';
export const basetoken_form_entertainment_02 = 'rgba(93, 101, 69, 0.4)';
export const basetoken_form_entertainment_03 = 'rgba(93, 101, 69, 0.24)';
export const basetoken_form_entertainment_04 = 'rgba(93, 101, 69, 0.08)';
export const basetoken_form_flower = '#486478';
export const basetoken_form_flower_01 = 'rgba(72, 100, 120, 0.64)';
export const basetoken_form_flower_02 = 'rgba(72, 100, 120, 0.4)';
export const basetoken_form_flower_03 = 'rgba(72, 100, 120, 0.24)';
export const basetoken_form_flower_04 = 'rgba(72, 100, 120, 0.08)';
export const basetoken_form_fog = '#4b635d';
export const basetoken_form_fog_01 = 'rgba(75, 99, 93, 0.64)';
export const basetoken_form_fog_02 = 'rgba(75, 99, 93, 0.4)';
export const basetoken_form_fog_03 = 'rgba(75, 99, 93, 0.24)';
export const basetoken_form_fog_04 = 'rgba(75, 99, 93, 0.08)';
export const basetoken_form_goal = '#997261';
export const basetoken_form_goal_01 = 'rgba(153, 114, 97, 0.64)';
export const basetoken_form_goal_02 = 'rgba(153, 114, 97, 0.4)';
export const basetoken_form_goal_03 = 'rgba(153, 114, 97, 0.24)';
export const basetoken_form_goal_04 = 'rgba(153, 114, 97, 0.08)';
export const basetoken_form_header_bg = 'rgba(222, 224, 227, 0.6)';
export const basetoken_form_header_bg_hover = 'rgba(222, 224, 227, 0.7)';
export const basetoken_form_header_bg_press = 'rgba(222, 224, 227, 0.8)';
export const basetoken_form_hourglass = '#535d89';
export const basetoken_form_hourglass_01 = 'rgba(83, 93, 137, 0.64)';
export const basetoken_form_hourglass_02 = 'rgba(83, 93, 137, 0.4)';
export const basetoken_form_hourglass_03 = 'rgba(83, 93, 137, 0.24)';
export const basetoken_form_hourglass_04 = 'rgba(83, 93, 137, 0.08)';
export const basetoken_form_input_bg = 'rgba(255, 255, 255, 0.08)';
export const basetoken_form_moon = '#494f73';
export const basetoken_form_moon_01 = 'rgba(73, 79, 115, 0.64)';
export const basetoken_form_moon_02 = 'rgba(73, 79, 115, 0.4)';
export const basetoken_form_moon_03 = 'rgba(73, 79, 115, 0.24)';
export const basetoken_form_moon_04 = 'rgba(73, 79, 115, 0.08)';
export const basetoken_form_peace = '#1c4cba';
export const basetoken_form_peace_01 = 'rgba(28, 76, 186, 0.64)';
export const basetoken_form_peace_02 = 'rgba(28, 76, 186, 0.4)';
export const basetoken_form_peace_03 = 'rgba(28, 76, 186, 0.24)';
export const basetoken_form_peace_04 = 'rgba(28, 76, 186, 0.08)';
export const basetoken_form_quiet = '#5e89a0';
export const basetoken_form_quiet_01 = 'rgba(94, 137, 160, 0.64)';
export const basetoken_form_quiet_02 = 'rgba(94, 137, 160, 0.4)';
export const basetoken_form_quiet_03 = 'rgba(94, 137, 160, 0.24)';
export const basetoken_form_quiet_04 = 'rgba(94, 137, 160, 0.08)';
export const basetoken_form_work = '#49608d';
export const basetoken_form_work_01 = 'rgba(73, 96, 141, 0.64)';
export const basetoken_form_work_02 = 'rgba(73, 96, 141, 0.4)';
export const basetoken_form_work_03 = 'rgba(73, 96, 141, 0.24)';
export const basetoken_form_work_04 = 'rgba(73, 96, 141, 0.08)';
export const basetoken_table_bg_blue_01 = '#f0f4ff';
export const basetoken_table_bg_blue_02 = '#e0e9ff';
export const basetoken_table_bg_blue_03 = '#c8d8ff';
export const basetoken_table_bg_blue_04 = '#a9c3ff';
export const basetoken_table_bg_carmine_01 = '#fef0f8';
export const basetoken_table_bg_carmine_02 = '#fee2f2';
export const basetoken_table_bg_carmine_03 = '#f9d0e7';
export const basetoken_table_bg_carmine_04 = '#f8b7db';
export const basetoken_table_bg_green_01 = '#e4fae1';
export const basetoken_table_bg_green_02 = '#d5f6d3';
export const basetoken_table_bg_green_03 = '#bfefc2';
export const basetoken_table_bg_green_04 = '#9de3a4';
export const basetoken_table_bg_lime_01 = '#f2f8d3';
export const basetoken_table_bg_lime_02 = '#ebf5bf';
export const basetoken_table_bg_lime_03 = '#e1eca7';
export const basetoken_table_bg_lime_04 = '#d0e085';
export const basetoken_table_bg_neutral_01 = '#f2f3f5';
export const basetoken_table_bg_neutral_02 = '#eff0f1';
export const basetoken_table_bg_neutral_03 = '#e5e6e9';
export const basetoken_table_bg_neutral_04 = '#d5d7da';
export const basetoken_table_bg_orange_01 = '#fff3e5';
export const basetoken_table_bg_orange_02 = '#fee7cd';
export const basetoken_table_bg_orange_03 = '#fed6ae';
export const basetoken_table_bg_orange_04 = '#ffc494';
export const basetoken_table_bg_purple_01 = '#f5f0fe';
export const basetoken_table_bg_purple_02 = '#efe6fe';
export const basetoken_table_bg_purple_03 = '#e3d4fd';
export const basetoken_table_bg_purple_04 = '#cfbaf3';
export const basetoken_table_bg_red_01 = '#fef0f0';
export const basetoken_table_bg_red_02 = '#fee3e2';
export const basetoken_table_bg_red_03 = '#fdd1d0';
export const basetoken_table_bg_red_04 = '#fabbb9';
export const basetoken_table_bg_turquoise_01 = '#e2f8f5';
export const basetoken_table_bg_turquoise_02 = '#caf3ee';
export const basetoken_table_bg_turquoise_03 = '#b7f3eb';
export const basetoken_table_bg_turquoise_04 = '#77ded1';
export const basetoken_table_bg_wathet_01 = '#e7f8fe';
export const basetoken_table_bg_wathet_02 = '#caeffc';
export const basetoken_table_bg_wathet_03 = '#b6e6fd';
export const basetoken_table_bg_wathet_04 = '#81d2f2';
export const basetoken_table_bg_yellow_01 = '#fbf4df';
export const basetoken_table_bg_yellow_02 = '#faedc2';
export const basetoken_table_bg_yellow_03 = '#fde9a5';
export const basetoken_table_bg_yellow_04 = '#fce088';
export const basetoken_table_details_process_bg = 'rgba(0, 0, 0, 0.03)';
