// @universe-design/styles/esm/base/index.dark.atom.js
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
export const basetoken_AI_instruction_hover = 'rgba(255, 255, 255, 0.1)';
export const basetoken_AI_instruction_normal = 'rgba(255, 255, 255, 0.05)';
export const basetoken_AI_instruction_press = 'rgba(255, 255, 255, 0.15)';
export const basetoken_container_bg = '#0a0a0a';
export const basetoken_dashboard_sidebar_bg_overlay = '#262626';
export const basetoken_dashboard_slicer_tag_bg_hover = 'rgba(10, 10, 10, 0.3)';
export const basetoken_dashboard_slicer_tag_bg_normal = 'rgba(10, 10, 10, 0.2)';
export const basetoken_form_award = '#909fb1';
export const basetoken_form_award_01 = 'rgba(144, 159, 177, 0.64)';
export const basetoken_form_award_02 = 'rgba(144, 159, 177, 0.4)';
export const basetoken_form_award_03 = 'rgba(144, 159, 177, 0.24)';
export const basetoken_form_award_04 = 'rgba(144, 159, 177, 0.08)';
export const basetoken_form_desert = '#aba69e';
export const basetoken_form_desert_01 = 'rgba(171, 166, 158, 0.64)';
export const basetoken_form_desert_02 = 'rgba(171, 166, 158, 0.4)';
export const basetoken_form_desert_03 = 'rgba(171, 166, 158, 0.24)';
export const basetoken_form_desert_04 = 'rgba(171, 166, 158, 0.08)';
export const basetoken_form_entertainment = '#afb2a4';
export const basetoken_form_entertainment_01 = 'rgba(175, 178, 164, 0.64)';
export const basetoken_form_entertainment_02 = 'rgba(175, 178, 164, 0.4)';
export const basetoken_form_entertainment_03 = 'rgba(175, 178, 164, 0.24)';
export const basetoken_form_entertainment_04 = 'rgba(175, 178, 164, 0.08)';
export const basetoken_form_flower = '#acbac1';
export const basetoken_form_flower_01 = 'rgba(172, 186, 193, 0.64)';
export const basetoken_form_flower_02 = 'rgba(172, 186, 193, 0.4)';
export const basetoken_form_flower_03 = 'rgba(172, 186, 193, 0.24)';
export const basetoken_form_flower_04 = 'rgba(172, 186, 193, 0.08)';
export const basetoken_form_fog = '#adb1ab';
export const basetoken_form_fog_01 = 'rgba(173, 177, 171, 0.64)';
export const basetoken_form_fog_02 = 'rgba(173, 177, 171, 0.4)';
export const basetoken_form_fog_03 = 'rgba(173, 177, 171, 0.24)';
export const basetoken_form_fog_04 = 'rgba(173, 177, 171, 0.08)';
export const basetoken_form_goal = '#a39281';
export const basetoken_form_goal_01 = 'rgba(163, 146, 129, 0.64)';
export const basetoken_form_goal_02 = 'rgba(163, 146, 129, 0.4)';
export const basetoken_form_goal_03 = 'rgba(163, 146, 129, 0.24)';
export const basetoken_form_goal_04 = 'rgba(163, 146, 129, 0.08)';
export const basetoken_form_header_bg = 'rgba(10, 10, 10, 0.4)';
export const basetoken_form_header_bg_hover = 'rgba(10, 10, 10, 0.5)';
export const basetoken_form_header_bg_press = 'rgba(10, 10, 10, 0.6)';
export const basetoken_form_hourglass = '#bab4a5';
export const basetoken_form_hourglass_01 = 'rgba(186, 180, 165, 0.64)';
export const basetoken_form_hourglass_02 = 'rgba(186, 180, 165, 0.4)';
export const basetoken_form_hourglass_03 = 'rgba(186, 180, 165, 0.24)';
export const basetoken_form_hourglass_04 = 'rgba(186, 180, 165, 0.08)';
export const basetoken_form_input_bg = 'rgba(68, 68, 68, 0.08)';
export const basetoken_form_moon = '#b6bbca';
export const basetoken_form_moon_01 = 'rgba(182, 187, 202, 0.64)';
export const basetoken_form_moon_02 = 'rgba(182, 187, 202, 0.4)';
export const basetoken_form_moon_03 = 'rgba(182, 187, 202, 0.24)';
export const basetoken_form_moon_04 = 'rgba(182, 187, 202, 0.08)';
export const basetoken_form_peace = '#9997b1';
export const basetoken_form_peace_01 = 'rgba(153, 151, 177, 0.64)';
export const basetoken_form_peace_02 = 'rgba(153, 151, 177, 0.4)';
export const basetoken_form_peace_03 = 'rgba(153, 151, 177, 0.24)';
export const basetoken_form_peace_04 = 'rgba(153, 151, 177, 0.08)';
export const basetoken_form_quiet = '#a6aeb9';
export const basetoken_form_quiet_01 = 'rgba(166, 174, 185, 0.64)';
export const basetoken_form_quiet_02 = 'rgba(166, 174, 185, 0.4)';
export const basetoken_form_quiet_03 = 'rgba(166, 174, 185, 0.24)';
export const basetoken_form_quiet_04 = 'rgba(166, 174, 185, 0.08)';
export const basetoken_form_work = '#beb9b1';
export const basetoken_form_work_01 = 'rgba(190, 185, 177, 0.64)';
export const basetoken_form_work_02 = 'rgba(190, 185, 177, 0.4)';
export const basetoken_form_work_03 = 'rgba(190, 185, 177, 0.24)';
export const basetoken_form_work_04 = 'rgba(190, 185, 177, 0.08)';
export const basetoken_table_bg_blue_01 = '#152340';
export const basetoken_table_bg_blue_02 = '#172f5e';
export const basetoken_table_bg_blue_03 = '#193a7c';
export const basetoken_table_bg_blue_04 = '#2655b6';
export const basetoken_table_bg_carmine_01 = '#321827';
export const basetoken_table_bg_carmine_02 = '#461b34';
export const basetoken_table_bg_carmine_03 = '#65284b';
export const basetoken_table_bg_carmine_04 = '#883564';
export const basetoken_table_bg_green_01 = '#0e2b0a';
export const basetoken_table_bg_green_02 = '#173b12';
export const basetoken_table_bg_green_03 = '#21511a';
export const basetoken_table_bg_green_04 = '#275e20';
export const basetoken_table_bg_lime_01 = '#212702';
export const basetoken_table_bg_lime_02 = '#303804';
export const basetoken_table_bg_lime_03 = '#404c06';
export const basetoken_table_bg_lime_04 = '#4d5a0c';
export const basetoken_table_bg_neutral_01 = '#292929';
export const basetoken_table_bg_neutral_02 = '#373737';
export const basetoken_table_bg_neutral_03 = '#4c4c4c';
export const basetoken_table_bg_neutral_04 = '#565656';
export const basetoken_table_bg_orange_01 = '#33210b';
export const basetoken_table_bg_orange_02 = '#4a2b10';
export const basetoken_table_bg_orange_03 = '#683a12';
export const basetoken_table_bg_orange_04 = '#7f4519';
export const basetoken_table_bg_purple_01 = '#281943';
export const basetoken_table_bg_purple_02 = '#381f61';
export const basetoken_table_bg_purple_03 = '#492688';
export const basetoken_table_bg_purple_04 = '#5c33a4';
export const basetoken_table_bg_red_01 = '#3d1a19';
export const basetoken_table_bg_red_02 = '#591f1d';
export const basetoken_table_bg_red_03 = '#7b2524';
export const basetoken_table_bg_red_04 = '#852e2c';
export const basetoken_table_bg_turquoise_01 = '#132926';
export const basetoken_table_bg_turquoise_02 = '#173b36';
export const basetoken_table_bg_turquoise_03 = '#1d4e47';
export const basetoken_table_bg_turquoise_04 = '#245e57';
export const basetoken_table_bg_wathet_01 = '#152830';
export const basetoken_table_bg_wathet_02 = '#103647';
export const basetoken_table_bg_wathet_03 = '#164359';
export const basetoken_table_bg_wathet_04 = '#135a78';
export const basetoken_table_bg_yellow_01 = '#30250a';
export const basetoken_table_bg_yellow_02 = '#473409';
export const basetoken_table_bg_yellow_03 = '#63470f';
export const basetoken_table_bg_yellow_04 = '#7f5d19';
export const basetoken_table_details_process_bg = 'rgba(255, 255, 255, 0.03)';
