// @universe-design/react/src/table/stories/bordered.stories.tsx
import React from 'react';
import type { TableProps, ColumnsType, ColumnType } from '../..';
import { Table } from '../..';
import type { Meta } from '@storybook/react';

import './index.less';
import { cssTypographyBody0, cssTypographyTitle3 } from '../../stylex';
import stylex from '@universe-design/stylex';

export default {
  title: 'Components/Table/临时示例',
  component: Table,
} satisfies Meta;

interface TableData {
  key: string;
  name: string;
  age: string;
  address: string;
  description: string;
  column1: string;
  column2: string;
  column3: string;
  children?: TableData[];
}

const columns = [
  {
    title: 'Name',
    dataIndex: 'name',
    width: '200px',
  },
  {
    title: 'Age',
    dataIndex: 'age',
  },
  {
    title: 'Address',
    dataIndex: 'address',
  },
  {
    title: 'Column 1',
    dataIndex: 'column1',
  },
  {
    title: 'Column 2',
    dataIndex: 'column2',
  },
  {
    title: 'Column 3',
    dataIndex: 'column3',
  },
  {
    title: 'Action',
    key: 'action',
    render: () => <a>Delete</a>,
  },
];

const generateData = (limit: number, level: number, prefix: number) => {
  const data: TableData[] = [];
  const levelStack = [];
  while (level) {
    levelStack.push(level);
    level--;
  }
  const levelPrefix = levelStack.join('-');
  for (let i = 1; i <= limit; i++) {
    data.push({
      key: `${prefix}-${levelPrefix}-${i}`,
      name: `John Brown ${levelPrefix}`,
      age: `${i}2`,
      address: `New York No. ${i} Lake Park`,
      description: `My name is John Brown, I am ${i}2 years old, living in New York No. ${i} Lake Park.`,
      column1: `New York No. ${i} Lake Park`,
      column2: `New York No. ${i} Lake Park`,
      column3: `New York No. ${i} Lake Park`,
    });
  }
  return data;
};

// 普通数据
const data = generateData(19, 1, 0);

export function BorderedTable() {
  const tableColumns: ColumnsType<TableData> = columns.map(item => ({ ...item } as ColumnType<TableData>));
  const state: TableProps<TableData> = {
    columns: tableColumns,
    dataSource: data,
  };

  const conditions: {
    description: string;
    proerty: keyof TableProps<TableData>;
    values: any[];
  }[] = [
    {
      description: 'whether has border ',
      proerty: 'bordered',
      values: [true, false],
    },
    {
      description: 'scrollY ',
      proerty: 'scroll',
      values: [{ y: 300 }, {}],
    },
    {
      description: 'scrollX ',
      proerty: 'scroll',
      values: [{ x: 600 }, {}],
    },
  ];

  type Sample = {
    description: string;
    tableProps: Partial<TableProps<TableData>>;
  };
  const samples: Sample[] = conditions.reduce((allSamples, condition) => {
    const newSamples: Sample[] = [];
    condition.values.forEach(conditionValue => {
      if (allSamples.length === 0) {
        newSamples.push({
          description: `${condition.description}: ${JSON.stringify(conditionValue)}`,
          tableProps: {
            [condition.proerty]: conditionValue,
          },
        });
      } else {
        allSamples.forEach(sample => {
          newSamples.push({
            description: `${sample.description}; ${condition.description}: ${JSON.stringify(conditionValue)}`,
            tableProps: {
              ...sample.tableProps,
              [condition.proerty]:
                typeof sample.tableProps[condition.proerty] === 'object'
                  ? Object.assign({}, sample.tableProps[condition.proerty], conditionValue)
                  : conditionValue,
            },
          });
        });
      }
    });
    return newSamples;
  }, [] as Sample[]);

  console.log(samples);
  return (
    <div style={{ padding: 40 }}>
      <div className={stylex(cssTypographyBody0)}>Enjoy everything about table border!</div>
      {samples.map((sample, index) => (
        <>
          <h3 className={stylex(cssTypographyTitle3)}>
            Case {index + 1}: {sample.description}
          </h3>
          <Table {...sample.tableProps} {...state} pagination={false} />
        </>
      ))}
    </div>
  );
}
