// @universe-design/react/src/table/stories/expand.stories.tsx
import React, { useState } from 'react';
import type { Meta } from '@storybook/react';
import type { TableProps, ColumnsType, ColumnType, DefaultSizeType } from '@universe-design/react';
import { Table, Radio, InputNumber, Button, Switch } from '@universe-design/react';

import './index.less';

export default {
  title: 'Components/Table/临时示例',
  component: Table,
} satisfies Meta;

interface TableData {
  key: string;
  name: string;
  age: string;
  address: string;
  description: string;
  column1: string;
  column2: string;
  column3: string;
  children?: TableData[];
}

const columns = [
  {
    title: 'Name',
    dataIndex: 'name',
    width: '200px',
  },
  {
    title: 'Age',
    dataIndex: 'age',
  },
  {
    title: 'Address',
    dataIndex: 'address',
  },
  {
    title: 'Column 1',
    dataIndex: 'column1',
  },
  {
    title: 'Column 2',
    dataIndex: 'column2',
  },
  {
    title: 'Column 3',
    dataIndex: 'column3',
  },
  {
    title: 'Action',
    key: 'action',
    render: () => (
      <Button type="text" color="primary">
        ActionBtn
      </Button>
    ),
  },
];

const generateData = (limit: number, level: number, prefix: number) => {
  const data: TableData[] = [];
  const levelStack = [];
  while (level) {
    levelStack.push(level);
    level--;
  }
  const levelPrefix = levelStack.join('-');
  for (let i = 1; i <= limit; i++) {
    data.push({
      key: `${prefix}-${levelPrefix}-${i}`,
      name: `John Brown ${levelPrefix}`,
      age: `${i}2`,
      address: `New York No. ${i} Lake Park`,
      description: `My name is John Brown, I am ${i}2 years old, living in New York No. ${i} Lake Park.`,
      column1: `New York No. ${i} Lake Park`,
      column2: `New York No. ${i} Lake Park`,
      column3: `New York No. ${i} Lake Park`,
    });
  }
  return data;
};

// 普通数据
const data = generateData(19, 1, 0);

export function ExpandbleTable() {
  const [size, setSize] = useState('medium' as DefaultSizeType);
  const [expandIconColumnIndex, setExpandIconColumnIndex] = useState(0);
  const [showSelection, setShowSelection] = useState(false);
  const tableColumns: ColumnsType<TableData> = columns.map(item => ({ ...item } as ColumnType<TableData>));
  const state: TableProps<TableData> = {
    columns: tableColumns,
    dataSource: data,
    size,
    expandable: {
      expandedRowRender: (record: TableData) => <p>{record.description}</p>,
      expandIconColumnIndex,
    },
  };

  return (
    <>
      <div className="actions">
        <div title="Size" className="action-item">
          Size:
          <Radio.Group value={size} onChange={v => setSize(v as DefaultSizeType)}>
            <Radio.Button value="large">Large</Radio.Button>
            <Radio.Button value="medium">medium</Radio.Button>
            <Radio.Button value="small">Small</Radio.Button>
          </Radio.Group>
          expandIconColumnIndex:{' '}
          <InputNumber value={expandIconColumnIndex} onChange={index => setExpandIconColumnIndex(index || 0)} />
          showSelection: <Switch checked={showSelection} onChange={checked => setShowSelection(checked)} />
        </div>
      </div>
      <Table
        {...state}
        pagination={false}
        scroll={{ x: 2000 }}
        columns={columns}
        rowSelection={
          showSelection
            ? {
                fixed: true,
              }
            : undefined
        }
        dataSource={data}
      />
    </>
  );
}
