// @universe-design/react/src/table/stories/fixcolumn.stories.tsx
import React from 'react';
import type { Meta } from '@storybook/react';
import type { TableProps, ColumnsType, ColumnType } from '../index';
import { Table } from '../index';

import './index.less';

export default {
  title: 'Components/Table/临时示例',
  component: Table,
} satisfies Meta;

interface TableData {
  key: string;
  name: string;
  age: string;
  address: string;
  description: string;
  column1: string;
  column2: string;
  column3: string;
  children?: TableData[];
}

const columns = [
  {
    title: 'Name',
    dataIndex: 'name',
    width: '200px',
  },
  {
    title: 'Age',
    dataIndex: 'age',
  },
  {
    title: 'Address',
    dataIndex: 'address',
  },
  {
    title: 'Column 1',
    dataIndex: 'column1',
  },
  {
    title: 'Column 2',
    dataIndex: 'column2',
  },
  {
    title: 'Column 3',
    dataIndex: 'column3',
  },
  {
    title: 'Action',
    key: 'action',
    render: () => <a>Delete</a>,
  },
];

const generateData = (limit: number, level: number, prefix: number) => {
  const data: TableData[] = [];
  const levelStack = [];
  while (level) {
    levelStack.push(level);
    level--;
  }
  const levelPrefix = levelStack.join('-');
  for (let i = 1; i <= limit; i++) {
    data.push({
      key: `${prefix}-${levelPrefix}-${i}`,
      name: `John Brown ${levelPrefix}`,
      age: `${i}2`,
      address: `New York No. ${i} Lake Park`,
      description: `My name is John Brown, I am ${i}2 years old, living in New York No. ${i} Lake Park.`,
      column1: `New York No. ${i} Lake Park`,
      column2: `New York No. ${i} Lake Park`,
      column3: `New York No. ${i} Lake Park`,
    });
  }
  return data;
};

// 普通数据
const data = generateData(19, 1, 0);

export function FixedColumnTable() {
  const tableColumns: ColumnsType<TableData> = columns.map(item => ({ ...item } as ColumnType<TableData>));
  tableColumns[0].fixed = true;
  tableColumns[tableColumns.length - 1].fixed = 'right';

  const state: TableProps<TableData> = {
    columns: tableColumns,
    dataSource: data,
  };

  return <Table {...state} scroll={{ x: '100vw', y: 240 }} pagination={false} />;
}
