// @universe-design/react/src/table/stories/homeless.stories.tsx
import React, { useState } from 'react';
import type { Meta } from '@storybook/react';
import type { TableProps, TablePaginationConfig, ColumnsType, ColumnType } from '../index';
import { Table } from '../index';
import type { DefaultSizeType } from '../../util';

import './index.less';
import type { SorterResult, Key, TableCurrentDataSource } from '../interface';
import { TableAction } from '../interface';
import { Space, Switch, Radio, Empty, Button, Tooltip } from '../..';
import { EmptyNeutralNoAccess } from '@universe-design/illustration-react';
import SucceedFilled from '@universe-design/icons-react/SucceedFilled';

export default {
  title: 'Components/Table/临时示例',
  component: Table,
} satisfies Meta;

interface TableData {
  key: string;
  name: string;
  age: string;
  address: string;
  description: string;
  column1: string;
  column2: string;
  column3: string;
  children?: TableData[];
}

const columns = [
  {
    title: 'Name',
    dataIndex: 'name',
    width: '200px',
  },
  {
    title: 'Age',
    dataIndex: 'age',
    sorter: (a: TableData, b: TableData) => parseInt(a.age) - parseInt(b.age),
  },
  {
    title: 'Address',
    dataIndex: 'address',
    filters: [
      {
        text: 'No. 4',
        value: 'No. 4',
      },
      {
        text: 'No. 8',
        value: 'No. 8',
      },
      {
        text: 'Submenu',
        value: 'Submenu',
        children: [
          {
            text: 'No. 12',
            value: 'No. 12',
          },
        ],
      },
    ],
    onFilter: (value: string, record: TableData) => record.address.includes(value),
  },
  {
    title: 'Column 1',
    dataIndex: 'column1',
  },
  {
    title: 'Column 2',
    dataIndex: 'column2',
  },
  {
    title: 'Column 3',
    dataIndex: 'column3',
  },
  {
    title: 'Action',
    key: 'action',
    render: () => <a>Delete</a>,
  },
];

const multiColumns = [
  {
    title: 'Name',
    dataIndex: 'name',
    width: '200px',
  },
  {
    title: 'Age',
    dataIndex: 'age',
    sorter: (a: TableData, b: TableData) => parseInt(a.age) - parseInt(b.age),
  },
  {
    title: 'Address',
    dataIndex: 'address',
    filters: [
      {
        text: 'London',
        value: 'London',
      },
      {
        text: 'New York',
        value: 'New York',
      },
    ],
    onFilter: (value: string, record: { address: string }) => record.address.indexOf(value) === 0,
  },
  {
    title: 'Column',
    children: [
      {
        title: 'Column 1',
        dataIndex: 'column1',
      },
      {
        title: 'Column 2',
        dataIndex: 'column2',
      },
      {
        title: 'Column 3',
        dataIndex: 'column3',
      },
    ],
  },
  {
    title: 'Action',
    key: 'action',
    render: () => <a>Delete</a>,
  },
];

const generateData = (limit: number, level: number, prefix: number) => {
  const data: TableData[] = [];
  const levelStack = [];
  while (level) {
    levelStack.push(level);
    level--;
  }
  const levelPrefix = levelStack.join('-');
  for (let i = 1; i <= limit; i++) {
    data.push({
      key: `${prefix}-${levelPrefix}-${i}`,
      name: `John Brown ${levelPrefix}`,
      age: `${i}2`,
      address: `New York No. ${i} Lake Park`,
      description: `My name is John Brown, I am ${i}2 years old, living in New York No. ${i} Lake Park.`,
      column1: `New York No. ${i} Lake Park`,
      column2: `New York No. ${i} Lake Park`,
      column3: `New York No. ${i} Lake Park`,
    });
  }
  return data;
};

// 普通数据
const data = generateData(19, 1, 0);

// 树形数据
const treeData = data.map(item => ({ ...item }));
[2, 4, 5, 7, 8].forEach(i => {
  const firstLevel = generateData(3, 2, i);
  const secondLevel = generateData(1, 3, i);
  firstLevel[1].children = secondLevel;
  treeData[i].children = firstLevel;
});

const expandableConfig = {
  expandedRowRender: (record: TableData) => <p>{record.description}</p>,
};
const titleConfig = () => 'Here is title';
const showHeaderConfig = true;
const footerConfig = () => 'Here is footer';
// const paginationConfig = { position: 'bottom' };

export function AllTable() {
  const [bordered, setBordered] = useState(false);
  const [loading, setLoading] = useState(false);
  // const [pagination, setPagination] = useState(paginationConfig);
  const [size, setSize] = useState('medium' as DefaultSizeType);
  const [expandable, setExpandable] = useState<any>();
  const [title, setTitle] = useState<any>();
  const [showHeader, setShowHeader] = useState(showHeaderConfig);
  const [showMultiHeader, setShowMultiHeader] = useState(false);
  const [footer, setFooter] = useState<any>();
  const [rowSelection, setRowSelection] = useState<any>();
  const [hasData, setHasData] = useState(true);
  const [tableLayout, setTableLayout] = useState(undefined);
  const [top, setTop] = useState('');
  const [bottom, setBottom] = useState('bottomRight');
  const [ellipsis, setEllipsis] = useState(false);
  const [xScroll, setxScroll] = useState('');
  const [yScroll, setyScroll] = useState(false);
  const [dataType, setDataType] = useState('normal');
  const [sticky, setSticky] = useState(false);
  const [checkStrictly, setCheckStrictly] = useState(true);

  const handleSizeChange = (v: any) => {
    setSize(v);
  };

  const handleTableLayoutChange = (e: any) => {
    setTableLayout(e);
  };

  const handleExpandChange = (enable: boolean) => {
    setExpandable(enable ? expandableConfig : null);
  };

  const handleEllipsisChange = (enable: boolean) => {
    setEllipsis(enable);
  };

  const handleTitleChange = (enable: boolean) => {
    setTitle(enable ? () => titleConfig : undefined);
  };

  const handleHeaderChange = (enable: boolean) => {
    setShowHeader(enable);
  };

  const handleFooterChange = (enable: boolean) => {
    setFooter(enable ? () => footerConfig : undefined);
  };

  const handleRowSelectionChange = (e: any) => {
    const type = e;
    setRowSelection(type ? { type } : undefined);
  };

  const handleYScrollChange = (enable: boolean) => {
    setyScroll(enable);
  };

  const handleXScrollChange = (e: any) => {
    setxScroll(e);
  };

  const handleDataChange = (enable: boolean) => {
    setHasData(enable);
  };

  const handlePaginationTop = (e: any) => {
    setTop(e);
  };

  const handlePaginationBottom = (e: any) => {
    setBottom(e);
  };

  const handleDataType = (e: any) => {
    setDataType(e);
  };

  const handleMultiHeaderChange = (enable: boolean) => {
    setShowMultiHeader(enable);
  };

  const handleStickyChange = (enable: boolean) => {
    setSticky(enable);
  };

  const handleCheckStrictlyChange = (enable: boolean) => {
    setCheckStrictly(enable);
  };

  const scroll: { x?: number | string; y?: number | string } = {};
  if (yScroll) {
    scroll.y = 240;
  }
  if (xScroll) {
    scroll.x = '100vw';
  }

  const tableColumns: ColumnsType<TableData> = columns.map(
    item => ({ ...item, ellipsis } as unknown as ColumnType<TableData>),
  );
  if (xScroll === 'fixed') {
    tableColumns[0].fixed = true;
    tableColumns[tableColumns.length - 1].fixed = 'right';
  }

  const multiTableColumns: ColumnsType<TableData> = multiColumns.map(
    item => ({ ...item, ellipsis } as unknown as ColumnType<TableData>),
  );
  if (xScroll === 'fixed') {
    multiTableColumns[0].fixed = true;
    multiTableColumns[multiTableColumns.length - 1].fixed = 'right';
  }

  const state: TableProps<TableData> = {
    columns: showMultiHeader ? multiTableColumns : tableColumns,
    dataSource: hasData ? (dataType === 'tree' ? treeData : data) : [],
    bordered,
    loading,
    showHeader,
    size,
    sticky,
  };

  if (expandable) {
    state.expandable = expandable;
  }
  if (title) {
    state.title = title;
  }
  if (footer) {
    state.footer = footer;
  }
  if (rowSelection) {
    state.rowSelection = rowSelection;
    if (state.rowSelection) {
      state.rowSelection.checkStrictly = checkStrictly;
    }
  }

  function onChange(
    pagination: TablePaginationConfig,
    filters: Record<string, (Key | boolean)[] | null>,
    sorter: SorterResult<TableData> | SorterResult<TableData>[],
    extra: TableCurrentDataSource<TableData>,
  ) {
    console.log('params', pagination, filters, sorter, extra);
  }

  return (
    <>
      <div className="actions">
        <div title="Bordered" className="action-item">
          Bordered: <Switch checked={bordered} onChange={setBordered} />
        </div>
        <div title="loading" className="action-item">
          Loading: <Switch checked={loading} onChange={setLoading} />
        </div>
        <div title="Title" className="action-item">
          Title: <Switch checked={Boolean(title)} onChange={handleTitleChange} />
        </div>
        <div title="Footer" className="action-item">
          Footer: <Switch checked={Boolean(footer)} onChange={handleFooterChange} />
        </div>
        <div title="Column Header" className="action-item">
          Column Header: <Switch checked={Boolean(showHeader)} onChange={handleHeaderChange} />
        </div>
        <div title="Multi Header" className="action-item">
          Multi Header: <Switch checked={Boolean(showMultiHeader)} onChange={handleMultiHeaderChange} />
        </div>
        <div title="Fixed Header" className="action-item">
          Fixed Header: <Switch checked={Boolean(yScroll)} onChange={handleYScrollChange} />
        </div>
        <div title="Sticky Header" className="action-item">
          Sticky Header: <Switch checked={Boolean(sticky)} onChange={handleStickyChange} />
        </div>
        <div title="Expandable" className="action-item">
          Expandable: <Switch checked={Boolean(expandable)} onChange={handleExpandChange} />
        </div>
        <div title="Ellipsis" className="action-item">
          Ellipsis: <Switch checked={Boolean(ellipsis)} onChange={handleEllipsisChange} />
        </div>
        <div title="CheckStrictly" className="action-item">
          CheckStrictly: <Switch checked={Boolean(checkStrictly)} onChange={handleCheckStrictlyChange} />
        </div>
        <div title="Row Selection" className="action-item">
          Row Selection:
          <Radio.Group value={rowSelection ? rowSelection.type : ''} onChange={handleRowSelectionChange}>
            <Radio.Button value="">Unset</Radio.Button>
            <Radio.Button value="radio">Radio</Radio.Button>
            <Radio.Button value="checkbox">Checkbox</Radio.Button>
          </Radio.Group>
        </div>
        <div title="Size" className="action-item">
          Size:
          <Radio.Group value={size} onChange={handleSizeChange}>
            <Radio.Button value="large">Large</Radio.Button>
            <Radio.Button value="medium">Medium</Radio.Button>
            <Radio.Button value="small">Small</Radio.Button>
          </Radio.Group>
        </div>
        <div title="Table Scroll" className="action-item">
          Table Scroll:
          <Radio.Group value={xScroll as any} onChange={handleXScrollChange}>
            <Radio.Button value="">Unset</Radio.Button>
            <Radio.Button value="scroll">Scroll</Radio.Button>
            <Radio.Button value="fixed">Fixed Columns</Radio.Button>
          </Radio.Group>
        </div>
        <div title="Table Layout" className="action-item">
          Table Layout:
          <Radio.Group value={tableLayout} onChange={handleTableLayoutChange}>
            <Radio.Button value="">Unset</Radio.Button>
            <Radio.Button value="fixed">Fixed</Radio.Button>
          </Radio.Group>
        </div>
        <div title="Pagination Top" className="action-item">
          Pagination Top:
          <Radio.Group value={top} onChange={handlePaginationTop}>
            <Radio.Button value="topLeft">TopLeft</Radio.Button>
            <Radio.Button value="topCenter">TopCenter</Radio.Button>
            <Radio.Button value="topRight">TopRight</Radio.Button>
            <Radio.Button value="">None</Radio.Button>
          </Radio.Group>
        </div>
        <div title="Pagination Bottom" className="action-item">
          Pagination Bottom:
          <Radio.Group value={bottom} onChange={handlePaginationBottom}>
            <Radio.Button value="bottomLeft">BottomLeft</Radio.Button>
            <Radio.Button value="bottomCenter">BottomCenter</Radio.Button>
            <Radio.Button value="bottomRight">BottomRight</Radio.Button>
            <Radio.Button value="">None</Radio.Button>
          </Radio.Group>
        </div>
        <div title="Has Data" className="action-item">
          Has Data: <Switch checked={Boolean(hasData)} onChange={handleDataChange} />
        </div>
        <div title="Pagination Bottom" className="action-item">
          Data:
          <Radio.Group value={dataType} onChange={handleDataType}>
            <Radio.Button value="normal">Normal</Radio.Button>
            <Radio.Button value="tree">Tree</Radio.Button>
          </Radio.Group>
        </div>
      </div>
      <Table
        {...state}
        tableLayout={tableLayout}
        scroll={scroll}
        pagination={!top && !bottom ? false : ({ position: [top, bottom] } as TablePaginationConfig)}
        onChange={onChange}
      />
    </>
  );
}

interface TableData2 {
  key: string;
  age: number;
  name: string;
  address: string;
}

const columns1: ColumnsType<TableData2> = [
  {
    title: 'Full Name',
    width: 100,
    dataIndex: 'name',
    key: 'name',
    fixed: 'left',
  },
  {
    title: 'Age',
    width: 100,
    dataIndex: 'age',
    key: 'age',
    fixed: 'left',
  },
  { title: 'Column 1', dataIndex: 'address', key: '1' },
  { title: 'Column 2', dataIndex: 'address', key: '2' },
  { title: 'Column 3', dataIndex: 'address', key: '3' },
  { title: 'Column 4', dataIndex: 'address', key: '4' },
  { title: 'Column 5', dataIndex: 'address', key: '5' },
  { title: 'Column 6', dataIndex: 'address', key: '6' },
  { title: 'Column 7', dataIndex: 'address', key: '7' },
  { title: 'Column 8', dataIndex: 'address', key: '8' },
  {
    title: 'Action',
    key: 'operation',
    fixed: 'right',
    width: 100,
    render: () => <a>Action</a>,
  },
];

const data1: TableData2[] = [
  {
    key: '1',
    name: 'John Brown',
    age: 32,
    address: 'New York Park',
  },
  {
    key: '2',
    name: 'Jim Green',
    age: 40,
    address: 'London Park',
  },
  {
    key: '3',
    name: 'Jim Green',
    age: 40,
    address: 'London Park',
  },
  {
    key: '4',
    name: 'Jim Green',
    age: 40,
    address: 'London Park',
  },
  {
    key: '5',
    name: 'Jim Green',
    age: 40,
    address: 'London Park',
  },
  {
    key: '6',
    name: 'Jim Green',
    age: 40,
    address: 'London Park',
  },
  {
    key: '7',
    name: 'Jim Green',
    age: 40,
    address: 'London Park',
  },
  {
    key: '8',
    name: 'Jim Green',
    age: 40,
    address: 'London Park',
  },
  {
    key: '9',
    name: 'Jim Green',
    age: 40,
    address: 'London Park',
  },
  {
    key: '10',
    name: 'Jim Green',
    age: 40,
    address: 'London Park',
  },
  {
    key: '11',
    name: 'Jim Green',
    age: 40,
    address: 'London Park',
  },
];

const columns3: ColumnsType<TableData2> = [
  {
    title: 'Full Name',
    width: 100,
    dataIndex: 'name',
    key: 'name',
    fixed: 'left',
  },
  {
    title: 'Age',
    width: 100,
    dataIndex: 'age',
    key: 'age',
    fixed: 'left',
  },
  {
    title: 'Action',
    key: 'operation',
    fixed: 'right',
    width: 100,
    render: () => <a>Action</a>,
  },
];

const columns5: ColumnsType<TableData2> = [
  {
    title: 'Full Name',
    width: 100,
    dataIndex: 'name',
    key: 'name',
  },
  {
    title: 'Age',
    width: 100,
    dataIndex: 'age',
    key: 'age',
  },
  { title: 'Column 1', dataIndex: 'address', key: '1' },
  { title: 'Column 2', dataIndex: 'address', key: '2' },
  { title: 'Column 3', dataIndex: 'address', key: '3' },
  { title: 'Column 4', dataIndex: 'address', key: '4' },
  { title: 'Column 5', dataIndex: 'address', key: '5' },
  { title: 'Column 6', dataIndex: 'address', key: '6' },
  { title: 'Column 7', dataIndex: 'address', key: '7' },
  { title: 'Column 8', dataIndex: 'address', key: '8' },
  {
    title: 'Action',
    key: 'operation',
    width: 100,
    render: () => <a>Action</a>,
  },
];

export const ScrollDemo = () => <Table columns={columns1} dataSource={data1} scroll={{ x: 1300, y: 200 }} sticky />;

export const HorizontalScrollDemo = () => <Table columns={columns5} dataSource={data1} scroll={{ x: 2400 }} />;

export const EmptyDemo = () => {
  const [bordered, setBordered] = useState<boolean>();
  const [size, setSize] = useState<DefaultSizeType>('medium');
  const [needColumns, setColumns] = useState<boolean>();
  return (
    <Space direction="vertical" style={{ width: '100%' }}>
      <Space>
        <div>
          <span>边框：</span>
          <Switch checked={bordered} onChange={setBordered} />
        </div>
        <div>
          <span>大小：</span>
          <Radio.Group value={size} onChange={v => setSize(v as DefaultSizeType)}>
            <Radio.Button value="large">大</Radio.Button>
            <Radio.Button value="medium">中</Radio.Button>
            <Radio.Button value="small">小</Radio.Button>
          </Radio.Group>
        </div>
        <div>
          <span>需要列配置：</span>
          <Switch checked={needColumns} onChange={setColumns} />
        </div>
      </Space>
      <Table bordered={bordered} size={size} dataSource={[]} columns={needColumns ? columns3 : []} pagination={false} />
      <Table
        bordered={bordered}
        size={size}
        dataSource={[]}
        columns={needColumns ? columns3 : []}
        pagination={false}
        renderEmpty="没数据了"
      />
      <Table
        bordered={bordered}
        size={size}
        dataSource={[]}
        columns={needColumns ? columns3 : []}
        renderEmpty={
          <Empty image={<EmptyNeutralNoAccess />} title="没有权限访问" description="请登录后尝试访问">
            <Button>返回主页</Button>
            <Button type="primary">申请权限</Button>
          </Empty>
        }
      />
    </Space>
  );
};

export const SortDemo = () => {
  const [ds, setDs] = React.useState<TableData2[]>();
  // const [pageSize, setPageSize] = React.useState<number>(10);
  const onChange = (_p: any, _f: any, s: any, a: TableCurrentDataSource<TableData2>) => {
    console.log(ds);
    if (a.action === TableAction.SORT) {
      console.log('currentDataSource', a.currentDataSource);
      const newData = [];
      for (let i = a.currentDataSource.length + 1; i <= a.currentDataSource.length + 10; i++) {
        newData.push({
          key: `${i}`,
          name: 'Jim Green',
          age: 40,
          address: 'London Park',
        });
      }
      setDs(ds?.concat(newData));
    }
  };

  React.useEffect(() => {
    const s = [];
    for (let i = 1; i <= 100; i++) {
      s.push({
        key: `${i}`,
        name: 'Jim Green',
        age: 40,
        address: 'London Park',
      });
    }
    setDs(s);
  }, []);
  return (
    <Table
      columns={columns1}
      dataSource={ds}
      pagination={{
        defaultPageSize: 5,
        // pageSize,
        pageSizeOptions: ['5', '10', '50'],
        onChange: (page, pageSizeState) => {
          console.log('pagination change', page, pageSizeState);
          // setPageSize(pageSizeState || 10);
        },
      }}
      onChange={onChange}
    />
  );
};

export const RowSelectionRenderCell = () => {
  const renderCell = (value: boolean, record: TableData2, index: number, originNode: React.ReactNode) => {
    console.log('renderCell', value, record, index);
    return <Tooltip title="tooltip demo">{originNode}</Tooltip>;
  };
  return (
    <Table
      scroll={{ x: '1000px' }}
      rowSelection={{
        renderCell,
      }}
      expandable={{
        expandedRowRender: record => <p>{record.address}</p>,
      }}
      columns={columns1}
      dataSource={data1}
    />
  );
};

const columns4 = [
  {
    title: 'Name',
    dataIndex: 'name',
  },
  {
    title: 'Salary',
    dataIndex: 'salary',
  },
  {
    title: 'Address',
    dataIndex: 'address',
  },
  {
    title: 'Email',
    dataIndex: 'email',
  },
];
const data4 = [
  {
    key: '1',
    name: 'Jane Doe',
    salary: 23000,
    address: '32 Park Road, London',
    email: 'jane.doe@example.com',
    children: [
      {
        key: '6',
        name: 'Alisa Ross',
        salary: 25000,
        address: '35 Park Road, London',
        email: 'alisa.ross@example.com',
      },
      {
        key: '7',
        name: 'Kevin Sandra',
        salary: 22000,
        address: '31 Park Road, London',
        email: 'kevin.sandra@example.com',
      },
    ],
  },
  // {
  //   key: '2',
  //   name: 'Alisa Ross',
  //   salary: 25000,
  //   address: '35 Park Road, London',
  //   email: 'alisa.ross@example.com',
  // },
  // {
  //   key: '3',
  //   name: 'Kevin Sandra',
  //   salary: 22000,
  //   address: '31 Park Road, London',
  //   email: 'kevin.sandra@example.com',
  //   children: [
  //     {
  //       key: '8',
  //       name: 'Ed Hellen',
  //       salary: 17000,
  //       address: '42 Park Road, London',
  //       email: 'ed.hellen@example.com',
  //     },
  //     {
  //       key: '9',
  //       name: 'William Smith',
  //       salary: 27000,
  //       address: '62 Park Road, London',
  //       email: 'william.smith@example.com',
  //     },
  //   ],
  // },
  // {
  //   key: '4',
  //   name: 'Ed Hellen',
  //   salary: 17000,
  //   address: '42 Park Road, London',
  //   email: 'ed.hellen@example.com',
  // },
  // {
  //   key: '5',
  //   name: 'William Smith',
  //   salary: 27000,
  //   address: '62 Park Road, London',
  //   email: 'william.smith@example.com',
  // },
];
export const customerExpandIcon = () => {
  const renderCell = (value: boolean, record: any, index: number, originNode: React.ReactNode) => {
    console.log('renderCell', value, record, index);
    return (
      <Tooltip title="tooltip demo">
        <div>{originNode}</div>
      </Tooltip>
    );
  };
  return (
    <Space
      direction="vertical"
      size="large"
      style={{
        width: '100%',
      }}
    >
      <Table
        columns={columns4}
        dataSource={data4}
        indentSize={80}
        rowSelection={{
          renderCell,
        }}
        expandable={{
          expandIcon: props => {
            const { onExpand, record, expandable } = props;
            if (!expandable) {
              return;
            }
            console.log('expandIcon', props);
            return (
              // 设计同学建议这里加个 tooltip, 方便查看 demo 的同学理解
              <Tooltip title="可以自定义展开 icon">
                <Button type="icon" icon={<SucceedFilled />} onClick={e => onExpand(record, e)} />
              </Tooltip>
            );
          },
        }}
      />
    </Space>
  );
};

export { default as StickyHeader } from './stickyHeader';
export { default as Summary } from './summary';
