// @universe-design/react/src/table/stories/preformance.stories.tsx
import React, { useState } from 'react';
import type { ColumnType } from '@universe-design/react';
import { Button, Table } from '@universe-design/react';
import type { Meta } from '@storybook/react';
const genRandom = (max: number) => Math.floor(Math.random() * max);

export default {
  title: 'Components/Table/临时示例',
  component: Table,
} satisfies Meta;

interface TableData {
  userInfo: string;
  indicatorUnit: number;
  startValue: number;
  targetValue: number;
  currentValue: number;
  children?: TableData[];
}

const initData = (
  count: number,
  params: {
    offset?: number;
    level2Count?: number;
    isTree?: boolean;
    level3Count?: number;
  },
) =>
  Array.from({ length: count }).map((_, _idx) => {
    let addon = 0;
    const uuid = Math.random().toString(36).slice(2);

    if (params.offset) {
      addon = params.offset;
    }

    const addonData = {} as any;

    if (params.isTree) {
      addonData.children = initData(params?.level2Count || 5, {
        offset: (_idx + 1) * 100,
      });
      addonData.children = addonData.children.map((child: any, index: number) => ({
        ...child,
        children: initData(params?.level3Count || 5, {
          offset: (_idx + 1) * 100 * 100 + (index + 1) * 100,
        }),
      }));
    }

    const idx = _idx + addon;

    return {
      id: uuid,
      key: idx,
      userInfo: `人员${idx + 1}`,
      indicatorUnit: idx,
      startValue: idx + genRandom(10),
      targetValue: idx,
      currentValue: idx,
      supportUserInfo: 'idx上级上级上级上级上级上级上级上级上级上级上级上级上级上级上级上级上级上级',
      associatedKr: Math.random() > 0.5 ? 'KR1' : 'KR2',
      action: idx,
      ...addonData,
    };
  });

const shouldCellUpdate = (pre: TableData, next: TableData) => pre !== next;
const sortColumns: ColumnType<TableData>[] = [
  {
    title: '人员',
    dataIndex: 'userInfo',
    width: 130,
    shouldCellUpdate,
  },
  {
    title: '单位',
    dataIndex: 'indicatorUnit',
    width: 110,
    render: () => '万元',
    shouldCellUpdate: (pre, next) => {
      console.log(pre, next);
      return false;
    },
  },
  {
    title: '起始值',
    dataIndex: 'startValue',
    width: 110,
    // 需要传入 sorter 才会生效... 有点诡异, 即便是纯数字也是如此
    sorter: {
      compare: (a: any, b: any) => a.startValue - b.startValue,
      multiple: 3,
    },
    shouldCellUpdate,
  },
  {
    title: '目标值',
    dataIndex: 'targetValue',
    width: 110,
    sorter: {
      compare: (a: any, b: any) => a.targetValue - b.targetValue,
      multiple: 1,
    },
    shouldCellUpdate,
  },
  {
    title: '进度值',
    dataIndex: 'currentValue',
    width: 164,
    shouldCellUpdate,
  },
  {
    title: '支撑的上级',
    dataIndex: 'supportUserInfo',
    width: 110,
    render: () => '上级',
    shouldCellUpdate,
  },
  {
    title: '关联 KR',
    dataIndex: 'associatedKr',
    width: 110,
    shouldCellUpdate,
  },
  {
    title: '操作',
    dataIndex: 'action',
    width: 110,
    render: () => '下发',
    shouldCellUpdate,
  },
];
const treeInitData = initData(1, {
  isTree: true,
  level2Count: 200,
  level3Count: 200,
});

export const TableMemo = () => {
  const [treeData, setTreeData] = useState(treeInitData);
  const moveIndex = React.useCallback(() => {
    const [{ children }] = treeData;
    const [firstChild, ...restChildren] = children;

    setTreeData([
      {
        ...treeData[0],
        children: [...restChildren, firstChild],
      },
    ]);
  }, [treeData]);
  return (
    <>
      <Button onClick={moveIndex}> moveIndex </Button>
      <Table columns={sortColumns} dataSource={treeData} />
    </>
  );
};
