// @universe-design/react/src/table/stories/screenshot.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import CEmptyLoadingScreenshot from '../demo/empty-loading.screenshot';
import CGreyhead from '../demo/greyhead';
import CHeadEllipsis from '../demo/head-ellipsis';
import CHeadLeftFilter from '../demo/head-left-filter';
import CNestedWithFixed from '../demo/nested-with-fixed';
import CStickyLoading from '../demo/sticky-loading';
import CStretch from '../demo/stretch';
if (!(CEmptyLoadingScreenshot as Story).storyName) {
  (CEmptyLoadingScreenshot as Story).storyName = 'Empty loading screenshot';
}
if (!(CGreyhead as Story).storyName) {
  (CGreyhead as Story).storyName = 'Greyhead';
}
if (!(CHeadEllipsis as Story).storyName) {
  (CHeadEllipsis as Story).storyName = 'Head ellipsis';
}
if (!(CHeadLeftFilter as Story).storyName) {
  (CHeadLeftFilter as Story).storyName = 'Head left filter';
}
if (!(CNestedWithFixed as Story).storyName) {
  (CNestedWithFixed as Story).storyName = 'Nested with fixed';
}
if (!(CStickyLoading as Story).storyName) {
  (CStickyLoading as Story).storyName = 'Sticky loading';
}
if (!(CStretch as Story).storyName) {
  (CStretch as Story).storyName = 'Stretch';
}

export default {
  title: 'Components/Table/截图测试',
  tags: ['autodocs'],
};

export {
  CEmptyLoadingScreenshot,
  CGreyhead,
  CHeadEllipsis,
  CHeadLeftFilter,
  CNestedWithFixed,
  CStickyLoading,
  CStretch,
};
