// @universe-design/react/src/table/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Basic from '../demo/0.basic';
import C1Select from '../demo/1.select';
import C2Expand from '../demo/2.expand';
import C4Nested from '../demo/4.nested';
import C5Tree from '../demo/5.tree';
import C6Head from '../demo/6.head';
import C7Virtualized from '../demo/7.virtualized';
import C8ResizableColumn from '../demo/8.resizable-column';
import CColspan from '../demo/colspan';
import CCustomFilter from '../demo/custom-filter';
import CEditCell from '../demo/edit-cell';
import CFixedColumns from '../demo/fixed-columns';
import CGroup from '../demo/group';
import CGroupingColumns from '../demo/grouping-columns';
import CProps from '../demo/props';
import CSticky from '../demo/sticky';
import CSummary from '../demo/summary';
if (!(C0Basic as Story).storyName) {
  (C0Basic as Story).storyName = '0. 基础用法';
  (C0Basic as Story).parameters = {
    docs: {
      description: {
        story: '基础表格。\n\n',
      },
    },
  };
}
if (!(C1Select as Story).storyName) {
  (C1Select as Story).storyName = '1. 选择模式';
  (C1Select as Story).parameters = {
    docs: {
      description: {
        story: '设置`rowSelection`的`type`属性可以开启选择模式。共有 2 个类型，`单选` 和 `多选`。\n\n',
      },
    },
  };
}
if (!(C2Expand as Story).storyName) {
  (C2Expand as Story).storyName = '2. 可展开';
  (C2Expand as Story).parameters = {
    docs: {
      description: {
        story: '通过`expandedRowRender`设置展开行。如果返回值是`null`，则不会渲染展开按钮。\n\n',
      },
    },
  };
}
if (!(C4Nested as Story).storyName) {
  (C4Nested as Story).storyName = '4. 嵌套子表格';
  (C4Nested as Story).parameters = {
    docs: {
      description: {
        story: '点击展开按钮可以在展开区域展示子表格。\n\n',
      },
    },
  };
}
if (!(C5Tree as Story).storyName) {
  (C5Tree as Story).storyName = '5. 树形数据展示';
  (C5Tree as Story).parameters = {
    docs: {
      description: {
        story: '当`data`中包含`children`或`childrenColumnName`时，会开启树形数据展示模式。\n\n',
      },
    },
  };
}
if (!(C6Head as Story).storyName) {
  (C6Head as Story).storyName = '6. 筛选和排序';
  (C6Head as Story).parameters = {
    docs: {
      description: {
        story:
          "对某一列数据进行筛选，使用列的`filters`属性来指定需要筛选菜单的列，`onFilter`用于筛选当前数据，`filterMultiple`用于指定多选和单选。\n\n对某一列数据进行排序，通过指定列的`sorter`函数即可启动排序按钮。`sorter: function(rowA, rowB) { ... }`， rowA、rowB 为比较的两个行数据。\n\n`sortDirections: ['ascend' | 'descend']`改变每列可用的排序方式，切换排序时按数组内容依次切换，设置在 Table Props 上时对所有列生效。你可以通过设置`['ascend', 'descend', 'ascend']`禁止排序恢复到默认状态。\n\n使用`defaultSortOrder`属性，设置列的默认排序顺序。\n\n",
      },
    },
  };
}
if (!(C7Virtualized as Story).storyName) {
  (C7Virtualized as Story).storyName = '7. 虚拟滚动';
  (C7Virtualized as Story).parameters = {
    docs: {
      description: {
        story: '通过 `react-window` 实现虚拟滚动表格。\n\n',
      },
    },
  };
}
if (!(C8ResizableColumn as Story).storyName) {
  (C8ResizableColumn as Story).storyName = '8. 可伸缩列';
  (C8ResizableColumn as Story).parameters = {
    docs: {
      description: {
        story: '集成 [`react-resizable`](https://github.com/react-grid-layout/react-resizable) 来实现可伸缩列。\n\n',
      },
    },
  };
}
if (!(CColspan as Story).storyName) {
  (CColspan as Story).storyName = '15. 合并单元格';
  (CColspan as Story).parameters = {
    docs: {
      description: {
        story:
          '表头只支持列合并，使用`column`里的`colSpan`进行设置。\n\n表格支持行 / 列合并，使用`render`里的单元格属性`colSpan`或者`rowSpan`。\n\n',
      },
    },
  };
}
if (!(CCustomFilter as Story).storyName) {
  (CCustomFilter as Story).storyName = '9. 自定义筛选菜单';
  (CCustomFilter as Story).parameters = {
    docs: {
      description: {
        story: '通过`filterDropdown`自定义列筛选功能，并实现一个搜索列的示例。\n\n',
      },
    },
  };
}
if (!(CEditCell as Story).storyName) {
  (CEditCell as Story).storyName = '23. 可编辑单元格';
  (CEditCell as Story).parameters = {
    docs: {
      description: {
        story: '带单元格编辑功能的表格。\n\n',
      },
    },
  };
}
if (!(CFixedColumns as Story).storyName) {
  (CFixedColumns as Story).storyName = '19. 固定列';
  (CFixedColumns as Story).parameters = {
    docs: {
      description: {
        story: '对于列数很多的数据，可以固定前后的列，横向滚动查看其它数据，需要和`scroll.x`配合使用。\n\n',
      },
    },
  };
}
if (!(CGroup as Story).storyName) {
  (CGroup as Story).storyName = '25. 表格数据分组';
  (CGroup as Story).parameters = {
    docs: {
      description: {
        story:
          '使用`groupBy`对表格数据进行分组，务必确保每个数据项有不同的 key。对于树型数据，分组仅对第一层数据有效\n\n',
      },
    },
  };
}
if (!(CGroupingColumns as Story).storyName) {
  (CGroupingColumns as Story).storyName = '22. 表头分组';
  (CGroupingColumns as Story).parameters = {
    docs: {
      description: {
        story: '`columns[n]`可以内嵌`children`，以渲染分组表头。\n\n',
      },
    },
  };
}
if (!(CProps as Story).storyName) {
  (CProps as Story).storyName = '99. 表格属性';
  (CProps as Story).parameters = {
    docs: {
      description: {
        story: '选择不同配置组合查看效果。\n\n',
      },
    },
  };
}
if (!(CSticky as Story).storyName) {
  (CSticky as Story).storyName = '10. 表头吸顶';
  (CSticky as Story).parameters = {
    docs: {
      description: {
        story: '通过设置`sticky`可以固定表头。\n\n',
      },
    },
  };
}
if (!(CSummary as Story).storyName) {
  (CSummary as Story).storyName = '24. 总结栏';
  (CSummary as Story).parameters = {
    docs: {
      description: {
        story:
          '总结栏可以使用 `Table.Summary` 同步 Column 的状态，你可以通过配置 `Table.Summary` 的 `fixed` 属性使其固定。`fixed` 共有四个值，\n\n- `top`，总结栏置顶\n- `bottom`，总结栏置底\n- true，同 `bottom`\n- false，不做置顶置底操作，跟随在 table 内容底部\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Table/官网示例',
  tags: ['autodocs'],
};

export {
  C0Basic,
  C1Select,
  C2Expand,
  C4Nested,
  C5Tree,
  C6Head,
  C7Virtualized,
  C8ResizableColumn,
  CColspan,
  CCustomFilter,
  CEditCell,
  CFixedColumns,
  CGroup,
  CGroupingColumns,
  CProps,
  CSticky,
  CSummary,
};
