// @universe-design/react/src/table/stories/stretch.stories.tsx
import React, { useMemo, useState } from 'react';
import type { Meta } from '@storybook/react';
import type { ColumnsType, ColumnType } from '../index';
import { Table } from '../index';
import { Divider, Empty, Radio, Space, Switch } from '../..';
import type { DefaultSizeType } from '../../util';
import { EmptyNegativeError } from '@universe-design/illustration-react';

import './index.less';

export default {
  title: 'Components/Table/临时示例',
  component: Table,
} satisfies Meta;

interface TableData {
  key: string;
  name: string;
  age: string;
  address: string;
  description: string;
  column1: string;
  column2: string;
  column3: string;
  children?: TableData[];
}

const columns = [
  {
    title: 'Name',
    dataIndex: 'name',
    width: '200px',
  },
  {
    title: 'Age',
    dataIndex: 'age',
  },
  {
    title: 'Address',
    dataIndex: 'address',
  },
  {
    title: 'Column 1',
    dataIndex: 'column1',
  },
  {
    title: 'Column 2',
    dataIndex: 'column2',
  },
  {
    title: 'Column 3',
    dataIndex: 'column3',
  },
  {
    title: 'Action',
    key: 'action',
    render: () => <a>Delete</a>,
  },
];

const generateData = (limit: number, level: number, prefix: number) => {
  const data: TableData[] = [];
  const levelStack = [];
  while (level) {
    levelStack.push(level);
    level--;
  }
  const levelPrefix = levelStack.join('-');
  for (let i = 1; i <= limit; i++) {
    data.push({
      key: `${prefix}-${levelPrefix}-${i}`,
      name: `John Brown ${levelPrefix}`,
      age: `${i}2`,
      address: `New York No. ${i} Lake Park`,
      description: `My name is John Brown, I am ${i}2 years old, living in New York No. ${i} Lake Park.`,
      column1: `New York No. ${i} Lake Park`,
      column2: `New York No. ${i} Lake Park`,
      column3: `New York No. ${i} Lake Park`,
    });
  }
  return data;
};

export function StretchTable() {
  const [size, setSize] = useState<DefaultSizeType>('medium');
  const [fixed, setFixed] = useState(false);
  const [rowCount, setRowCount] = useState(3);

  const tableColumns: ColumnsType<TableData> = columns.map(item => ({ ...item } as ColumnType<TableData>));

  const tableData = useMemo<TableData[]>(() => generateData(rowCount, 1, 0), [rowCount]);

  if (fixed) {
    tableColumns[0].fixed = 'left';
    tableColumns[tableColumns.length - 1].fixed = 'right';

    for (let i = 0; i < tableColumns.length; i++) {
      tableColumns[i].width = 300;
    }
  }

  return (
    <Space direction="vertical" style={{ width: '100%' }} split={<Divider dashed />}>
      <Space>
        <div>
          <span>大小：</span>
          <Radio.Group value={size} onChange={v => setSize(v as DefaultSizeType)}>
            <Radio.Button value="large">大</Radio.Button>
            <Radio.Button value="medium">中</Radio.Button>
            <Radio.Button value="small">小</Radio.Button>
          </Radio.Group>
        </div>
        <div>
          <span>数据量：</span>
          <Radio.Group value={rowCount} onChange={v => setRowCount(v as number)}>
            <Radio.Button value={3}>3 条</Radio.Button>
            <Radio.Button value={19}>19 条</Radio.Button>
            <Radio.Button value={22}>22 条</Radio.Button>
          </Radio.Group>
        </div>
        <div>
          <span>Fixed：</span>
          <Switch checked={fixed} onChange={setFixed} />
        </div>
      </Space>
      <Table
        stretch
        style={{ height: 500 }}
        size={size}
        columns={tableColumns}
        dataSource={tableData}
        renderEmpty={<Empty image={<EmptyNegativeError />} title="居中显示" />}
      />
      <Table
        stretch
        style={{ height: 500 }}
        size={size}
        columns={tableColumns}
        dataSource={[]}
        pagination={{ current: 3, total: 1000 }}
      />
      <Table
        stretch
        style={{ height: 500 }}
        size={size}
        columns={tableColumns}
        dataSource={[]}
        pagination={{
          current: 3,
          total: 1000,
          position: ['bottomCenter'],
          size: 'small',
        }}
      />
    </Space>
  );
}
