// @universe-design/react/src/table/stories/summary.tsx
import React from 'react';
import { Divider, Table } from '@universe-design/react';

const columns = [
  {
    title: 'Name',
    dataIndex: 'name',
    fixed: true,
  },
  {
    title: 'Borrow',
    dataIndex: 'borrow',
  },
  {
    title: 'Repayment',
    dataIndex: 'repayment',
  },
];

const data: any[] = [];
for (let i = 0; i < 20; i += 1) {
  data.push({
    key: i,
    name: ['John Brown', 'Jim Green', 'Joe Black'][i % 3],
    borrow: 10 + 3 * i,
    repayment: 20 + i,
  });
}

const fixedColumns = [
  {
    title: 'Name',
    dataIndex: 'name',
    fixed: true,
    width: 100,
  },
  {
    title: 'Description',
    dataIndex: 'description',
  },
];

const fixedData: any[] = [];
for (let i = 0; i < 8; i += 1) {
  fixedData.push({
    key: i,
    name: ['Light', 'Bamboo', 'Little'][i % 3],
    description: 'Everything that has a beginning, has an end.',
  });
}

const Demos = ({ sticky, summaryFixed }: { sticky: boolean; summaryFixed: 'top' | 'bottom' | false }) => (
  <>
    <Divider>
      Sticky: {JSON.stringify(sticky)}, Summary Fixed: {JSON.stringify(summaryFixed)}
    </Divider>
    <Table
      sticky={sticky}
      columns={columns}
      dataSource={data}
      pagination={false}
      scroll={{ y: 400 }}
      bordered
      summary={pageData => {
        let totalBorrow = 0;
        let totalRepayment = 0;

        pageData.forEach(({ borrow, repayment }) => {
          totalBorrow += borrow;
          totalRepayment += repayment;
        });

        return (
          <Table.Summary fixed={summaryFixed}>
            <Table.Summary.Row>
              <Table.Summary.Cell index={0}>Total</Table.Summary.Cell>
              <Table.Summary.Cell index={1}>
                <span>{totalBorrow}</span>
              </Table.Summary.Cell>
              <Table.Summary.Cell index={2}>
                <span>{totalRepayment}</span>
              </Table.Summary.Cell>
            </Table.Summary.Row>
            <Table.Summary.Row>
              <Table.Summary.Cell index={0}>Balance</Table.Summary.Cell>
              <Table.Summary.Cell index={1} colSpan={2}>
                <span>{totalBorrow - totalRepayment}</span>
              </Table.Summary.Cell>
            </Table.Summary.Row>
          </Table.Summary>
        );
      }}
    />

    <br />

    <Table
      sticky={sticky}
      columns={fixedColumns}
      dataSource={fixedData}
      pagination={false}
      scroll={{ x: 2000 }}
      bordered={true}
      summary={() => (
        <Table.Summary fixed={summaryFixed}>
          <Table.Summary.Row>
            <Table.Summary.Cell index={0}>Summary</Table.Summary.Cell>
            <Table.Summary.Cell index={1}>This is a summary content</Table.Summary.Cell>
          </Table.Summary.Row>
        </Table.Summary>
      )}
    />
  </>
);

export default () => (
  <div style={{ margin: '0 180px' }}>
    <div style={{ height: 2000 }} />
    <Demos sticky summaryFixed="top" />
    <Demos sticky summaryFixed="bottom" />
    <Demos sticky summaryFixed={false} />
    <Demos sticky={false} summaryFixed="top" />
    <Demos sticky={false} summaryFixed="bottom" />
    <Demos sticky={false} summaryFixed={false} />
    <div style={{ height: 2000 }} />
  </div>
);
