// @universe-design/react/lib/dialog/dialog/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useDialogLayoutMode = exports.DialogLayoutMode = exports.DialogCancelType = exports.Dialog = void 0;
var _ukchsq = _interopRequireDefault(require("./../../.stylex/ukchsq7"));
var _uqt5uie = _interopRequireDefault(require("./../../.stylex/uqt5uie"));
var _uudhrvd = _interopRequireDefault(require("./../../.stylex/uudhrvd"));
var _u16y10k = _interopRequireDefault(require("./../../.stylex/u16y10k6"));
var _ua38rkj = _interopRequireDefault(require("./../../.stylex/ua38rkj"));
var _ulh0qmm = _interopRequireDefault(require("./../../.stylex/ulh0qmm"));
var _u16ivsb = _interopRequireDefault(require("./../../.stylex/u16ivsb9"));
var _uup8z = _interopRequireDefault(require("./../../.stylex/uup8z65"));
var _uizwrz = _interopRequireDefault(require("./../../.stylex/uizwrz6"));
var _ujgul0d = _interopRequireDefault(require("./../../.stylex/ujgul0d"));
var _u1981n0b = _interopRequireDefault(require("./../../.stylex/u1981n0b"));
var _ug0oh3b = _interopRequireDefault(require("./../../.stylex/ug0oh3b"));
var _u135sb2c = _interopRequireDefault(require("./../../.stylex/u135sb2c"));
var _u1uvgrag = _interopRequireDefault(require("./../../.stylex/u1uvgrag"));
var _u5jxv = _interopRequireDefault(require("./../../.stylex/u5jxv71"));
var _u1p85sbs = _interopRequireDefault(require("./../../.stylex/u1p85sbs"));
var _ul98dlq = _interopRequireDefault(require("./../../.stylex/ul98dlq"));
var _u57ipeq = _interopRequireDefault(require("./../../.stylex/u57ipeq"));
var _umejf3v = _interopRequireDefault(require("./../../.stylex/umejf3v"));
var _u1o4axus = _interopRequireDefault(require("./../../.stylex/u1o4axus"));
var _uc0fbx = _interopRequireDefault(require("./../../.stylex/uc0fbx2"));
var _u11ie86f = _interopRequireDefault(require("./../../.stylex/u11ie86f"));
var _u1ust6hi = _interopRequireDefault(require("./../../.stylex/u1ust6hi"));
var _uugkpzq = _interopRequireDefault(require("./../../.stylex/uugkpzq"));
var _uif4zit = _interopRequireDefault(require("./../../.stylex/uif4zit"));
var _u1mcpqdr = _interopRequireDefault(require("./../../.stylex/u1mcpqdr"));
var _uq34mcg = _interopRequireDefault(require("./../../.stylex/uq34mcg"));
var _u5txyzo = _interopRequireDefault(require("./../../.stylex/u5txyzo"));
var _u4w7uz = _interopRequireDefault(require("./../../.stylex/u4w7uz9"));
var _u175h1m = _interopRequireDefault(require("./../../.stylex/u175h1m2"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _clsx = _interopRequireDefault(require("clsx"));
var React = _interopRequireWildcard(require("react"));
var _reactRemoveScrollBar = require("react-remove-scroll-bar");
var _reactTransitionGroup = require("react-transition-group");
var _ConfigReducedMotion = require("../../config-provider/ConfigReducedMotion");
var _reactProvider = require("@universe-design/react-provider");
var _hooks = require("../../config-provider/hooks");
var _hooks2 = require("../../hooks");
var _portal = _interopRequireDefault(require("../../portal"));
var _shared = require("../../shared");
var _useAnimatedVisible = require("../../shared/components/transition/useAnimatedVisible");
var _util = require("../../util");
var _composeRefs = _interopRequireDefault(require("../../util/composeRefs"));
var _forwardRef = require("../../util/forwardRef");
var _reactUtilities = require("@universe-design/react-utilities");
require("../style");
//.u22iagw{display:flex;}
//.u6jr5hl{align-items:flex-start;}
//.u1aehjj5{pointer-events:none;}
//.utm89h2{max-width:1080px;}
//.u7wyk6f{min-width:420px;}
//.u1rcxdsg{width:calc(100vw - 32px * 2);}
//.u1g96bv0{max-width:840px;}
//.u177j64f{width:600px;}
//.u1j052i9{min-height:128px;}
//.u1qzxk2a{width:420px;}
//.u113wtx2{top:0px;}
//.utuwxu6{display:inline-flex;}
//.u1o700av{text-align:left;}
//.umrv4ls{vertical-align:middle;}
//.u10pi13n{position:relative;}
//.u1ewtqcl{box-sizing:border-box;}
//.u4djv7l{top:100px;}
//.u1hu3pq6{margin-top:0;}
//.u1ujusj6{margin-right:auto;}
//.u19f4twv{margin-bottom:0;}
//.ucgxt0o{margin-left:auto;}
//.u6juhto{white-space:normal;}
//.u17mccla{text-align:center;}
//.uz5stix{white-space:nowrap;}
//.ud1hv5k::before{display:inline-block;}
//.ugovu44::before{width:0px;}
//.ur6rhvx::before{height:100%;}
//.u1o5b37y::before{vertical-align:middle;}
//.uyl8oag::before{box-sizing:border-box;}
//.uzhtfnv::before{content:'';}
/* eslint-disable max-nested-callbacks */
/**
 * @deprecated
 */
const {
  ConfigProvider,
  useConfig: useDialogLayoutMode,
  setGlobal
} = /* #__PURE__*/(0, _reactProvider.createGlobalConfig)('dialog-layout-mode', '2.18.0');
exports.useDialogLayoutMode = useDialogLayoutMode;
/**
 * @deprecated
 * 为了更好支持设计规范，我们在 `2.18.0` 修改了对话框的布局方式，由于修改布局方式，对于之前覆盖过对话框样式的用户可能产生 breaking change。默认情况下，我们会使用 `2.18.0` 新布局，如果用户侧存量代码较多，验证成本较高，可以使用该组件修改子树中的对话框布局模式为 `legacy`。
 *
 * 建议渐进式验证并替换为新版 layout，本组件将会在下个大版本中移除
 */
const DialogLayoutMode = exports.DialogLayoutMode = Object.assign(function (_ref) {
  let {
    mode,
    children
  } = _ref;
  return /*#__PURE__*/React.createElement(ConfigProvider, {
    value: mode
  }, children);
}, {
  setGlobal
});
let DialogCancelType = exports.DialogCancelType = /*#__PURE__*/function (DialogCancelType) {
  DialogCancelType[DialogCancelType["CancelButton"] = 0] = "CancelButton";
  DialogCancelType[DialogCancelType["CloseIcon"] = 1] = "CloseIcon";
  DialogCancelType[DialogCancelType["Esc"] = 2] = "Esc";
  DialogCancelType[DialogCancelType["Mask"] = 3] = "Mask";
  return DialogCancelType;
}({});
function useDialogCssScope() {
  return (0, _hooks.useCssScope)('dialog');
}
const defaultSafeMarginHmin = 32;
const defaults = {
  getContainer: (0, _reactUtilities.isClient)() ? () => document.body : false,
  safeMargin: {
    top: 32,
    bottom: 32,
    hmax: 180,
    hmin: defaultSafeMarginHmin
  }
};
const defaultSizeCfg = {
  small: {
    minWidth: 420,
    maxWidth: 420
  },
  medium: {
    minWidth: 600,
    maxWidth: 600
  },
  large: {
    minWidth: 420,
    maxWidth: 840
  },
  'extra-large': {
    minWidth: 420,
    maxWidth: 1080
  }
};
const DialogContent = (0, _forwardRef.forwardHTMLRef)((_ref2, ref) => {
  let {
    centered,
    safeMargin,
    size,
    width: widthProp,
    minWidth: minWidthProp,
    maxWidth: maxWidthProp,
    style: styleProp = {},
    in: inProp,
    ...restProps
  } = _ref2;
  // safeMargin
  const {
    top: safeMarginTop = defaults.safeMargin.top,
    bottom: safeMarginBottom = defaults.safeMargin.bottom,
    hmax: safeMarginHMax = defaults.safeMargin.hmax,
    hmin: safeMarginHMin = defaults.safeMargin.hmin
  } = safeMargin;
  const contentRef = React.useRef(null);
  const mergedContentRef = React.useMemo(() => (0, _composeRefs.default)(ref, contentRef), [ref]);

  /**
   * 调整 dialog content 的 top left width height，用于自适应安全边距和最大最小宽度高度。
   * **注意**: 这里 height 相关的实现和技术设计中的描述有一些出入，如果要获取准确的 minHeight 和 maxHeight 需要修改容器的 maxHeight 值再获取 scrollHeight，但是这个过程有两个问题：
   * 1. 会引起 force reflow，会有性能问题 https://gist.github.com/paulirish/5d52fb081b3570c81e3a
   * 2. scrollTop 会失效，需要 reset 后代节点的 scrollTop
   *
   * 为了避免不必要的性能损耗，这里尽可能通过 style 状态和父容器状态进行 nextStyle 的推算，给节点做 nextStyle 赋值，从而走浏览器自己的 batch update 流程。有必要的话可以把这个方法暴露给用户。
   */
  const adjustContentInset = React.useCallback(function () {
    var _contentEl$parentElem, _containerRect$width, _containerRect$height, _ref3, _ref4;
    const contentEl = contentRef.current;
    if (!contentEl) {
      return;
    }
    if (!contentEl.offsetParent) {
      return;
    }
    const computedStyle = getComputedStyle(contentEl);

    // minWidth & maxWidth
    // 优先级 minWidthProp > style.minWidth > computedStyle.minWidth > widthProp > style.width > computedStyle.width > size config
    const minWidth = (() => {
      if ((0, _util.isNonNullable)(minWidthProp)) {
        return minWidthProp;
      }
      if ((0, _util.isNonNullable)(styleProp.minWidth)) {
        return styleProp.minWidth;
      }
      if (computedStyle.minWidth !== '0px' && !contentEl.style.minWidth) {
        // css minWidth
        return computedStyle.minWidth;
      }
    })();
    const maxWidth = (() => {
      if ((0, _util.isNonNullable)(maxWidthProp)) {
        return maxWidthProp;
      }
      if ((0, _util.isNonNullable)(styleProp.maxWidth)) {
        return styleProp.maxWidth;
      }
      if (computedStyle.maxWidth !== 'none' && !contentEl.style.maxWidth) {
        return computedStyle.maxWidth;
      }
    })();
    const win = (0, _util.getWindow)(contentEl);
    // safari 下 scrollHeight 不准所以用 getBoundingClientRect
    const containerRect = (_contentEl$parentElem = contentEl.parentElement) === null || _contentEl$parentElem === void 0 ? void 0 : _contentEl$parentElem.getBoundingClientRect();
    const containerWidth = (_containerRect$width = containerRect === null || containerRect === void 0 ? void 0 : containerRect.width) !== null && _containerRect$width !== void 0 ? _containerRect$width : win.innerWidth;
    const containerHeight = (_containerRect$height = containerRect === null || containerRect === void 0 ? void 0 : containerRect.height) !== null && _containerRect$height !== void 0 ? _containerRect$height : win.innerHeight;
    const _width = (() => {
      if ((0, _util.isNonNullable)(widthProp)) {
        return widthProp;
      }
      if ((0, _util.isNonNullable)(styleProp.width)) {
        return styleProp.width;
      }
    })();
    const minWidthNum = (0, _util.parseLength)((_ref3 = minWidth !== null && minWidth !== void 0 ? minWidth : _width) !== null && _ref3 !== void 0 ? _ref3 : defaultSizeCfg[size].minWidth, containerWidth);
    const maxWidthNum = (0, _util.parseLength)((_ref4 = maxWidth !== null && maxWidth !== void 0 ? maxWidth : _width) !== null && _ref4 !== void 0 ? _ref4 : defaultSizeCfg[size].maxWidth, containerWidth);

    /**
     * 处理宽度
     */
    const {
      width,
      marginH
    } = function () {
      // width 算出来可能是一个非整数，我们不做处理，不然可能会产生 width 抖动导致 reflow
      if (containerWidth < minWidthNum + safeMarginHMin * 2) {
        return {
          width: `${minWidthNum}px`,
          marginH: `${safeMarginHMin}px`
        };
      }
      if (containerWidth < minWidthNum + safeMarginHMax * 2) {
        const margin = (containerWidth - minWidthNum) / 2;
        return {
          width: `${containerWidth - margin * 2}px`,
          marginH: `${margin}px`
        };
      }
      if (containerWidth < maxWidthNum + safeMarginHMax * 2) {
        return {
          width: `${containerWidth - safeMarginHMax * 2}px`,
          marginH: `${safeMarginHMax}px`
        };
      }
      const margin = (containerWidth - maxWidthNum) / 2;
      return {
        width: `${containerWidth - margin * 2}px`,
        marginH: `${margin}px`
      };
    }();

    /**
     * 处理高度
     */
    const {
      height,
      marginTop
    } = function () {
      const _top = (() => {
        if ((0, _util.isNonNullable)(styleProp.top)) {
          return styleProp.top;
        }
        if (computedStyle.top !== '0px' && !contentEl.style.top) {
          return computedStyle.top;
        }
        return centered ? 0 : 100;
      })();
      const contentTop = (0, _util.parseLength)(_top, containerHeight);
      if (centered) {
        const margin = Math.max(safeMarginTop, safeMarginBottom);
        return {
          height: `${containerHeight - margin * 2}px`,
          marginTop: `${margin}px`
        };
      }
      if (contentTop < safeMarginTop) {
        return {
          height: `${containerHeight - safeMarginTop - safeMarginBottom}px`,
          marginTop: `${safeMarginTop}px`
        };
      }
      return {
        height: `${containerHeight - contentTop - safeMarginBottom}px`,
        marginTop: `${contentTop}px`
      };
    }();

    // 用 marginRight 为了在 wrapper 水平滚动的时候右侧可以撑开，但是很不幸只有最新版的 chrome 是这个行为，其他版本 margin 都是不能占位的。
    // 不用 padding 有两个原因，一是 padding 会导致 mask 无法点击到，需要特殊处理，二是会跟之前的实现有些出入，可能导致用户挂掉
    // TODO marginBottom 在竖直滚动的时候也无法撑开
    contentEl.style.width = width;
    contentEl.style.marginLeft = marginH;
    contentEl.style.marginRight = marginH;
    contentEl.style.height = height;
    contentEl.style.top = marginTop;
  }, [widthProp, styleProp.width, styleProp.minWidth, styleProp.maxWidth, styleProp.top, minWidthProp, size, maxWidthProp, safeMarginHMin, safeMarginHMax, centered, safeMarginTop, safeMarginBottom]);
  (0, _hooks2.useIsomorphicLayoutEffect)(() => {
    if (inProp) {
      adjustContentInset();
    }
  }, [adjustContentInset, inProp]);
  React.useEffect(() => {
    const win = (0, _util.getWindow)(contentRef.current);
    const cb = () => adjustContentInset();
    win.addEventListener('resize', cb);
    return () => {
      win.removeEventListener('resize', cb);
    };
  });
  return /*#__PURE__*/React.createElement("div", (0, _extends2.default)({
    ref: mergedContentRef,
    style: styleProp
  }, restProps));
});
const DialogContentLegacy = (0, _forwardRef.forwardHTMLRef)((_ref5, ref) => {
  let {
    centered,
    safeMargin,
    ...restProps
  } = _ref5;
  process.env.NODE_ENV !== "production" ? (0, _util.warning)(false, (0, _util.devWarning)('Dialog',
  // eslint-disable-next-line max-len
  "You are using `Dialog` in legacy mode, legacy mode has been deprecated and will be removed in the next major release. The following properties cannot be used: `minWidth` `maxWidth` `safeMargin.hmin` `safeMargin.hmax` `maxHeight` `bodyHeightResponsiveMode` in legacy mode. Please remove <DialogLayoutMode mode='legacy' /> or wrap your <Dialog /> with <DialogLayoutMode mode='2.18.0' /> for better support for design specs.")) : void 0;
  const {
    top: safeMarginTop = defaults.safeMargin.top,
    bottom: safeMarginBottom = defaults.safeMargin.bottom
  } = safeMargin;
  const contentRef = React.useRef(null);
  React.useImperativeHandle(ref, () => contentRef.current, []);
  (0, _hooks2.useIsomorphicLayoutEffect)(() => {
    // 矫正 content top 不超过安全边距
    if (contentRef.current) {
      let contentTop = parseFloat(getComputedStyle(contentRef.current).top || '0');
      if (!centered && contentTop < safeMarginTop) {
        contentTop = safeMarginTop;
        contentRef.current.style.top = `${safeMarginTop}px`;
      } else if (centered && contentTop > 0) {
        // top 属性失效
        contentTop = 0;
        contentRef.current.style.top = '0px';
      }
      contentRef.current.style.maxHeight = `calc(100vh - ${centered ? safeMarginTop + safeMarginBottom : safeMarginBottom + contentTop}px)`;
    }
  });
  return /*#__PURE__*/React.createElement("div", (0, _extends2.default)({
    ref: contentRef
  }, restProps));
});

// eslint-disable-next-line max-lines-per-function
const Dialog = exports.Dialog = /*#__PURE__*/React.forwardRef((props, ref) => {
  const {
    visible = false,
    overlay = true,
    overlayClosable,
    keyboardClosable = true,
    contentStyle,
    overlayStyle,
    overlayClassName,
    contentClassName,
    wrapperClassName,
    enableAnimation: enableAnimationProp,
    overlayTransitionClassName,
    contentTransitionClassName,
    destroyOnClose,
    children,
    width,
    zIndex,
    centered = true,
    getContainer = defaults.getContainer,
    onClose,
    onAfterClose,
    onAfterOpen,
    size = 'medium',
    safeMargin = defaults.safeMargin,
    className,
    initialFocusRef,
    finalFocusRef,
    enableFocusLock = false,
    enableRestoreFocus = false,
    forceRender,
    minWidth,
    maxWidth,
    ...restProps
  } = props;
  const dialogCssScope = useDialogCssScope();
  const layoutMode = useDialogLayoutMode();
  const enableAnimation = !(0, _ConfigReducedMotion.unstable_useConfigReducedMotion)((0, _util.isNonNullable)(enableAnimationProp) ? !enableAnimationProp : undefined);
  const wrapperRef = React.useRef(null);
  const contentRef = React.useRef(null);
  const contentClickRef = React.useRef();
  const contentTimeoutRef = React.useRef();

  // 添加 esc 监听
  const onKeyDown = React.useCallback(e => {
    if (keyboardClosable) {
      const {
        keyCode
      } = e;
      /* istanbul ignore else  */
      if (keyCode === _util.KeyCode.ESC) {
        onClose && onClose(e, {
          cancelType: DialogCancelType.Esc
        });
      }
    }
  }, [keyboardClosable, onClose]);
  const handleOverlayClick = e => {
    if (!overlayClosable) {
      return;
    }
    if (contentClickRef.current) {
      contentClickRef.current = false;
    } else if (wrapperRef.current === e.target) {
      onClose && onClose(e, {
        cancelType: DialogCancelType.Mask
      });
    }
  };
  const onContentMouseDown = () => {
    window.clearTimeout(contentTimeoutRef.current);
    contentClickRef.current = true;
  };
  const onContentMouseUp = () => {
    contentTimeoutRef.current = window.setTimeout(() => {
      contentClickRef.current = false;
    });
  };
  const {
    transitionGroupProps,
    transitionGroupPropsWithoutListener,
    styleDisplay,
    noopRef
  } = (0, _useAnimatedVisible.useAnimatedVisible)({
    destroyOnClose,
    forceRender,
    visible,
    onAfterClose,
    onAfterOpen
  });
  const timeout = React.useMemo(() => ({
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    enter: enableAnimation ? 400 : 0,
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    exit: enableAnimation ? 300 : 0
  }), [enableAnimation]);
  const maskRef = React.useRef(null);
  const mergedInitialFocusRef = initialFocusRef === null ? null : initialFocusRef !== null && initialFocusRef !== void 0 ? initialFocusRef : wrapperRef;
  const renderOverlay = () => {
    if (!overlay) {
      return null;
    }
    const cls = (0, _clsx.default)(dialogCssScope('__mask'), overlayClassName);
    let animationCls = '';
    if (enableAnimation) {
      animationCls = overlayTransitionClassName || dialogCssScope('__mask-animation');
    }
    const mergeStyle = {
      ...overlayStyle,
      ...styleDisplay
    };
    if (zIndex || zIndex === 0) {
      mergeStyle.zIndex = zIndex;
    }
    return /*#__PURE__*/React.createElement(_reactTransitionGroup.CSSTransition, (0, _extends2.default)({
      timeout: timeout,
      appear: true,
      classNames: animationCls,
      nodeRef: maskRef
    }, transitionGroupPropsWithoutListener), /*#__PURE__*/React.createElement("div", {
      ref: maskRef,
      style: mergeStyle,
      className: cls
    }));
  };
  const usedSize = (0, _util.getSizeAbbr)(size);
  const renderContent = () => {
    const wrapCls = (0, _clsx.default)(dialogCssScope('__wrap'), wrapperClassName, {
      [dialogCssScope('--centered')]: centered
    }, (0, _stylex.default)(layoutMode === 'legacy' ? [centered && {
      "props": {
        "fsow6f": _u1ust6hi.default,
        "Huce71": _uugkpzq.default,
        "Br0sdwz": _uif4zit.default,
        "Ccq8qp": _u1mcpqdr.default,
        "Baz25je": _uq34mcg.default,
        "Byamso": _u5txyzo.default,
        "Bcvre1j": _u4w7uz.default,
        "Ftih45": _u175h1m.default
      }
    }] : [centered && {
      "props": {
        "Huce71": _uugkpzq.default
      }
    }]));
    const cls = (0, _clsx.default)(dialogCssScope('__content'), contentClassName, {
      [dialogCssScope(`__content-${usedSize}`)]: usedSize
    }, (0, _stylex.default)(layoutMode === 'legacy' ? [{
      "props": {
        "qhf8xq": _u5jxv.default,
        "B7ck84d": _u1p85sbs.default,
        "Bhzewxz": _ul98dlq.default,
        "B6of3ja": _u57ipeq.default,
        "t21cq0": _umejf3v.default,
        "jrapky": _u1o4axus.default,
        "Frg6f3": _uc0fbx.default,
        "Huce71": _u11ie86f.default
      }
    }, centered && {
      "props": {
        "Bhzewxz": _u1981n0b.default,
        "mc9l5x": _ug0oh3b.default,
        "fsow6f": _u135sb2c.default,
        "ha4doy": _u1uvgrag.default
      }
    }, usedSize && {
      sm: {
        "props": {
          "a9b677": _ujgul0d.default
        }
      },
      md: {
        "props": {
          "a9b677": _uup8z.default,
          "sshi5w": _uizwrz.default
        }
      },
      lg: {
        "props": {
          "Bf4jedk": _ua38rkj.default,
          "a9b677": _ulh0qmm.default,
          "B2u0y6b": _u16ivsb.default
        }
      },
      xl: {
        "props": {
          "Bf4jedk": _ua38rkj.default,
          "a9b677": _ulh0qmm.default,
          "B2u0y6b": _u16y10k.default
        }
      }
    }[usedSize]] : [{
      "props": {
        "qhf8xq": _u5jxv.default,
        "Huce71": _u11ie86f.default,
        "mc9l5x": _ukchsq.default,
        "Bt984gj": _uqt5uie.default,
        "Bkecrkj": _uudhrvd.default
      }
    }]));
    const wrapStyle = {
      ...styleDisplay
    };
    if (zIndex || zIndex === 0) {
      wrapStyle.zIndex = zIndex;
    }
    let animationCls = '';
    if (enableAnimation) {
      animationCls = contentTransitionClassName || dialogCssScope('__content-animation');
    }
    return /*#__PURE__*/React.createElement(_reactTransitionGroup.CSSTransition, (0, _extends2.default)({
      timeout: timeout,
      appear: true,
      classNames: animationCls,
      nodeRef: contentRef // 这里改为 contentRef 因为在用 ScrollArea 的时候和之前用 div 的时候行为不同，关闭 dialog 会有偶现的闪动
    }, transitionGroupPropsWithoutListener), /*#__PURE__*/React.createElement(_shared.FocusLock, {
      isDisabled: !visible,
      initialFocusRef: mergedInitialFocusRef,
      restoreFocus: enableRestoreFocus,
      finalFocusRef: finalFocusRef,
      enableFocusLock: enableFocusLock
    }, /*#__PURE__*/React.createElement("div", {
      className: wrapCls,
      style: wrapStyle,
      tabIndex: -1,
      onClick: handleOverlayClick,
      onKeyDown: onKeyDown,
      ref: wrapperRef
    }, layoutMode === 'legacy' ? /*#__PURE__*/React.createElement(DialogContentLegacy, {
      style: {
        ...contentStyle,
        width
      },
      className: cls,
      onMouseDown: onContentMouseDown,
      onMouseUp: onContentMouseUp,
      safeMargin: safeMargin,
      centered: centered,
      role: "dialog",
      "aria-modal": "true",
      ref: contentRef
    }, children) : /*#__PURE__*/React.createElement(DialogContent, {
      style: contentStyle,
      className: cls,
      onMouseDown: onContentMouseDown,
      onMouseUp: onContentMouseUp,
      safeMargin: safeMargin,
      centered: centered,
      role: "dialog",
      "aria-modal": "true",
      ref: contentRef,
      size: size,
      width: width,
      minWidth: minWidth,
      maxWidth: maxWidth,
      in: transitionGroupPropsWithoutListener.in
    }, children))));
  };
  return /*#__PURE__*/React.createElement(_reactTransitionGroup.Transition, (0, _extends2.default)({
    timeout: timeout,
    appear: true,
    nodeRef: noopRef
  }, transitionGroupProps), (() => {
    const dialogRootNode = /*#__PURE__*/React.createElement("div", (0, _extends2.default)({
      ref: ref,
      className: (0, _clsx.default)(dialogCssScope('__root'), className)
    }, restProps), visible && /*#__PURE__*/React.createElement(_reactRemoveScrollBar.RemoveScrollBar, {
      noRelative: true,
      gapMode: "margin"
    }), renderOverlay(), renderContent());
    if (getContainer === false) {
      return dialogRootNode;
    }
    return /*#__PURE__*/React.createElement(_portal.default, {
      container: getContainer
    }, dialogRootNode);
  })());
});
