// @universe-design/react/src/config-provider/index.tsx
import * as React from 'react';
import { ThemeProvider as UDIconThemeProvider } from '@universe-design/icons-react/ThemeProvider';
import { PopupPortalContainerProvider, usePopupPortalContainer } from '@universe-design/react-provider';
import { useContext } from 'use-context-selector';
import { useSyncExternalStore } from 'use-sync-external-store/shim';
import mergeWith from 'lodash/mergeWith';
import uniq from 'lodash/uniq';
import type {
  ConfigProviderContextType,
  ConfigProviderProps,
  DirectionType,
  TextOverflowPopperSidecarProps,
} from './interface';
import { configStore } from './miniStore';
import { isClient } from '@universe-design/react-utilities';
import { useRenderEffect } from '../hooks/useRenderEffect';
import { ConfigContext } from './ConfigContext';
import { defaultConfig } from './constants';

import './style';

export * from './desktop';

// ConfigProvider，几种情况
// 1. CrossRootConfigProvider 或者 ConfigProvider 如果有父级，则直接用父级和 props merge 得到当前层的结果
// 2. CrossRootConfigProvider 如果没有父级，则使用 module scope 的 config 作为父级
// 3. ConfigProvider 如果没有父级，则需要将当前层级的 config 更新到 module scope config

/**
 * `ConfigProvider` 可嵌套，props 可传递，用于解决 light 中配置 dark 的情况。
 * `useLocaleFontFamily`和对`miniStore`操作全局唯一。
 */

interface ConfigChildrenProps extends Omit<ConfigProviderProps, 'getPopupContainer'> {
  parentContext: ConfigProviderContextType;
  isOutermost: boolean;
}

const ConfigChildren: React.FC<ConfigChildrenProps> = props => {
  const {
    parentContext = defaultConfig,
    isOutermost,
    // props
    locale = parentContext.locale,
    prefixCls = parentContext.rootPrefixCls,
    direction = parentContext.direction,
    getTargetContainer = parentContext.getTargetContainer,
    toast = parentContext.toast,
    theme = parentContext.theme,
    textOverflowPopperSidecar = parentContext.textOverflowPopperSidecar,
    useLocaleFontFamily = false,
    componentSize = parentContext.componentSize,
    children,
    PopperSidecar = parentContext.PopperSidecar,
    timezone = parentContext.timezone,
  } = props;

  // 处理 partial locale
  const finalLocale = React.useMemo(
    () =>
      // 递归合并 defaultLocale 和 locale，但遇到数组时只进行浅覆盖
      mergeWith({}, parentContext.locale, locale, (objValue, srcValue) => {
        if (Array.isArray(objValue)) {
          return srcValue;
        }
      }),
    [locale, parentContext.locale],
  );

  const getPopupContainer = usePopupPortalContainer();
  const config = React.useMemo(
    () => ({
      rootPrefixCls: prefixCls,
      locale: finalLocale,
      direction,
      getPopupContainer,
      getTargetContainer,
      toast,
      theme,
      componentSize,
      textOverflowPopperSidecar,
      PopperSidecar,
      timezone,
    }),
    [
      prefixCls,
      finalLocale,
      direction,
      getPopupContainer,
      getTargetContainer,
      toast,
      theme,
      componentSize,
      textOverflowPopperSidecar,
      PopperSidecar,
      timezone,
    ],
  );

  useRenderEffect(
    React.useCallback(() => {
      if (!isOutermost || !isClient()) {
        return;
      }

      if (!(useLocaleFontFamily && locale.locale)) {
        return;
      }

      const lang = locale.locale.split('-')[0];

      if (!lang) {
        return;
      }

      const localeFontClassName = `${prefixCls}__lang-${lang}`;

      const getClassNames = () => document.body.className.split(/\s+/);
      document.body.className = uniq([...getClassNames(), localeFontClassName]).join(' ');
      return () => {
        document.body.className = getClassNames()
          .filter(o => o !== localeFontClassName)
          .join(' ');
      };
    }, [isOutermost, locale.locale, prefixCls, useLocaleFontFamily]),
  );

  React.useEffect(() => {
    if (!isOutermost) {
      return;
    }
    configStore.set(config);
  }, [isOutermost, config]);

  return (
    <UDIconThemeProvider mode={config.theme}>
      <ConfigContext.Provider value={config}>{children}</ConfigContext.Provider>
    </UDIconThemeProvider>
  );
};

/**
 * 内部使用
 * @internal
 */
export const CrossRootConfigProvider: React.FC<React.PropsWithChildren<unknown>> = props => {
  const { children } = props;
  const globalConfig = useSyncExternalStore(configStore.subscribe, configStore.get, configStore.get);

  const child = (
    <ConfigChildren parentContext={globalConfig} isOutermost={false}>
      {children}
    </ConfigChildren>
  );

  if (globalConfig.getPopupContainer) {
    return <PopupPortalContainerProvider value={globalConfig.getPopupContainer}>{child}</PopupPortalContainerProvider>;
  }
  return child;
};

/**
 *
 * <CrossRootConfigProvider>
 *   <ConfigChildren />
 * </CrossRootConfigProvider>
 *
 * <ConfigProvider>
 *   <ConfigChildren />
 * </ConfigProvider>
 */
export const ConfigProvider: React.FC<ConfigProviderProps> = props => {
  const parentContext = useContext(ConfigContext);
  const { getPopupContainer, ...restProp } = props;
  const isOutermost = parentContext === defaultConfig;
  const child = <ConfigChildren parentContext={parentContext} isOutermost={isOutermost} {...restProp} />;

  if (getPopupContainer) {
    return <PopupPortalContainerProvider value={getPopupContainer}>{child}</PopupPortalContainerProvider>;
  }
  return child;
};

export type ConsumerConfig = ConfigProviderContextType;

export interface ConsumerProps<T = ConsumerConfig> {
  children: (config: T) => React.ReactElement;
}

export const Consumer: React.FC<ConsumerProps> = props => {
  const config = useContext(ConfigContext);
  return props.children({ ...config });
};

export const COMPONENT_VERSION_ID = '__VERSION__';

export type { ConfigProviderProps, DirectionType, TextOverflowPopperSidecarProps };

export {
  useComponentSize,
  useInputGroupSize,
  useCssScope,
  useLocale,
  useClassPrefix,
  useLocaleReceiver,
  useDirection,
  useGetTargetContainer,
  useTheme,
} from './hooks';

export { ConfigContext };
