// @universe-design/react/src/tree/index.tsx
import * as React from 'react';
import { COMPONENT_VERSION_ID } from '../config-provider';
import type { TreeProps } from './interface';
import { ShowCheckedStrategy } from './interface';
import { TreeNode } from './TreeNode';
import { TreeContextProvider } from './hooks/TreeContext';
import type { TreeRef } from './hooks/useTree';
import { useTree } from './hooks/useTree';
import { TreeContainer } from './TreeContainer';

import './style';

export const Tree = Object.assign(
  React.forwardRef<TreeRef, TreeProps>(function Tree(props, ref) {
    const treeCtxValue = useTree(props, ref);
    return (
      <TreeContextProvider value={treeCtxValue}>
        <TreeContainer />
      </TreeContextProvider>
    );
  }),
  {
    SHOW_ALL: ShowCheckedStrategy.SHOW_ALL,
    SHOW_PARENT: ShowCheckedStrategy.SHOW_PARENT,
    SHOW_CHILD: ShowCheckedStrategy.SHOW_CHILD,
    TreeNode,
    [COMPONENT_VERSION_ID]: process.env.__VERSION__,
  },
);

export * from './interface';
export * from './util';
export type { TreeRef, TreeListItemReference } from './hooks/useTree';
export * from './hooks/useTreeCheck';
export * from './hooks/useTreeSelect';
export * from './TreeNode';
export * from './TreeNodeCheckbox';
export * from './TreeNodeCheckmark';
export * from './TreeNodeContainer';
export * from './TreeNodeExpandButton';
export * from './TreeNodeIndents';
export * from './TreeNodeLabel';
