// @universe-design/react/src/tree/TreeNodeCheckbox.tsx
import React from 'react';
import clsx from 'clsx';
import { useTreeContext } from './hooks/TreeContext';
import { CheckedType } from './interface';
import type { CheckboxProps } from '../checkbox';
import { Checkbox } from '../checkbox/Checkbox';
import { useTreeNodeContext } from './hooks/TreeNodeContext';
import { isCheckedDisabled } from './util';
import composeRefs from '../util/composeRefs';
import { isFunction } from '../util';

export const TreeNodeCheckbox = React.forwardRef<
  HTMLInputElement,
  Omit<CheckboxProps, 'defaultChecked' | 'checked' | 'indeterminate'>
>(
  (
    {
      // to be merged
      className,
      onChange,
      // to be omitted
      disabled,
      ...restProps
    },
    ref,
  ) => {
    const { checkable, toggleChecked, rootCssScope } = useTreeContext();
    const { doublyTreeNode: node, checkboxRef, isChecked } = useTreeNodeContext();

    /**
     * 合并 contextRef 和 userRef
     */
    const mergedRef = React.useMemo(() => composeRefs(checkboxRef, ref), [checkboxRef, ref]);

    const { value } = node.__node__;
    const checkedDisabled = isCheckedDisabled(node.__node__);

    let checkboxOverlapProps: CheckboxProps = {
      disabled: disabled ?? checkedDisabled,
      className: clsx(rootCssScope('__node__checkbox'), className),
      onChange: React.useCallback<NonNullable<CheckboxProps['onChange']>>(
        e => {
          isFunction(onChange) && onChange(e);
          toggleChecked(value);
        },
        [onChange, toggleChecked, value],
      ),
    };
    if (isChecked === CheckedType.Checked) {
      checkboxOverlapProps = {
        checked: true,
        indeterminate: false,
        ...checkboxOverlapProps,
      };
    } else if (isChecked === CheckedType.Unchecked) {
      checkboxOverlapProps = {
        checked: false,
        indeterminate: false,
        ...checkboxOverlapProps,
      };
    } else if (isChecked === CheckedType.Indeterminate) {
      checkboxOverlapProps = {
        checked: false,
        indeterminate: true,
        ...checkboxOverlapProps,
      };
    }

    if (checkable) {
      return <Checkbox ref={mergedRef} {...checkboxOverlapProps} {...restProps} />;
    }

    return null;
  },
);
