// @universe-design/react/src/tree/util.ts
/* eslint-disable max-depth */
import type { TKeyValueMap } from '../util';
import { devWarning, warning } from '../util';
import { Queue } from '../util/datastructure/Queue';
import type { LeafTreeNodeConfig, NonLeafTreeNodeConfig, TreeConfig, TreeNodeConfig } from './interface';

export function cloneMap<K, V>(map: Map<K, V>): Map<K, V> {
  const newMap = new Map<K, V>();
  for (const key of map.keys()) {
    const v = map.get(key);
    v !== undefined && newMap.set(key, v);
  }
  return newMap;
}

/**
 * 遍历 tree config
 * @param tree
 * @param cb
 */
export function traverseTreeConfig<T extends { children?: T[] }>(
  tree: T[],
  cb: (node: T, index: number, depth: number, parent: T | null) => void,
) {
  let index = 0;
  function _traverse(_node: T, _parentDepth: number, _parent: T | null) {
    const _depth = _parentDepth + 1;
    if (!_node.children) {
      /* empty */
    }
    cb(_node, index, _depth, _parent);
    index += 1;
    for (const __node of _node.children || []) {
      _traverse(__node, _depth, _node);
    }
  }
  // 深度优先遍历
  for (const node of tree) {
    _traverse(node, 0, null);
  }
}

export function array2Set(arr: string[]): TKeyValueMap<boolean> {
  const res: TKeyValueMap<boolean> = {};
  arr.forEach(o => (res[o] = true));
  return res;
}

export function array2Map<T>(arr: T[], callbackfn: (item: T) => string): TKeyValueMap<T> {
  const res: TKeyValueMap<T> = {};
  for (const item of arr) {
    const k = callbackfn(item);

    // 如果检测到重复的 key, 给出警告
    warning(res[k] === undefined, devWarning('tree', `Array element is repeated, ${k} already exists`));

    res[k] = item;
  }
  return res;
}

export function map2Array(map: TKeyValueMap<boolean>): string[] {
  return Object.keys(map);
}

export function isLeaf(node: TreeNodeConfig): node is LeafTreeNodeConfig {
  return !node.children || node.children.length === 0;
}

export function isNonLeaf(node: TreeNodeConfig): node is NonLeafTreeNodeConfig {
  return Boolean(node.children && node.children.length !== 0);
}

export function isCheckedDisabled(node: TreeNodeConfig): boolean | undefined {
  return node.disableCheckbox || node.disabled;
}

/**
 * @public
 */
export interface DoublyTreeNode {
  /**
   * key
   */
  key: string;
  /**
   * 当前节点详情
   */
  __node__: TreeNodeConfig;
  /**
   * 父节点
   */
  __parent__: TreeNodeConfig | null;
  /**
   * 层级
   */
  depth: number;
  /**
   * 父节点
   */
  parent: DoublyTreeNode | null;
  /**
   * 子节点
   */
  children?: DoublyTreeNode[];
}

export function getTreeNodeConfigPath(doublyTreeNode: DoublyTreeNode | undefined | null) {
  const res: TreeNodeConfig[] = [];
  let childDoublyTreeNode = doublyTreeNode;
  while (childDoublyTreeNode) {
    res.push(childDoublyTreeNode.__node__);
    childDoublyTreeNode = childDoublyTreeNode.parent;
  }
  return res;
}

export function isNonLeaf2(node: DoublyTreeNode): node is DoublyTreeNode & Required<Pick<DoublyTreeNode, 'children'>> {
  return Boolean(node.children && node.children.length !== 0);
}

/**
 * 生成一个双向链表 map
 * @param tree
 */
export function genDoublyTreeNodeMap(tree: TreeConfig): {
  doublyTreeNodeMap: TKeyValueMap<DoublyTreeNode>;
  doublyTreeNodeList: DoublyTreeNode[];
} {
  const doublyTreeNodeList: DoublyTreeNode[] = [];
  const doublyTreeNodeWeakMap = new WeakMap<TreeNodeConfig, DoublyTreeNode>();
  traverseTreeConfig(tree, (node, index, depth, parent) => {
    const wrappedNode = {
      key: node.value,
      __node__: node,
      __parent__: parent,
      depth,
      parent: null,
    };
    doublyTreeNodeList.push(wrappedNode);
    doublyTreeNodeWeakMap.set(node, wrappedNode);
  });

  const doublyTreeNodeMap: TKeyValueMap<DoublyTreeNode> = array2Map(doublyTreeNodeList, o => o.__node__.value);
  doublyTreeNodeList.forEach(node => {
    node.parent = node.__parent__ ? doublyTreeNodeWeakMap.get(node.__parent__) || null : null;
    node.children = node.__node__.children?.map(c => doublyTreeNodeWeakMap.get(c)!);
  });
  return { doublyTreeNodeMap, doublyTreeNodeList };
}

export function traverseDoublyTreeNode(node: DoublyTreeNode, cb: (node: DoublyTreeNode) => void) {
  for (let stack: DoublyTreeNode[] = [node]; ; ) {
    const current = stack.pop();
    if (!current) {
      break;
    }
    cb(current);
    current.children?.forEach(o => stack.push(o));
  }
}

export function getTreeNodesAncestors(
  doublyTreeNodes: DoublyTreeNode[],
  filterTreeNode: (node: DoublyTreeNode) => boolean,
) {
  const nodeSet = new Set<DoublyTreeNode>();
  doublyTreeNodes.forEach(node => {
    if (nodeSet.has(node)) {
      return;
    }
    // 如果命中则所有的祖先节点全部加入
    if (filterTreeNode(node)) {
      // 祖先节点
      for (let current: DoublyTreeNode | null = node.parent; ; ) {
        if (!current || nodeSet.has(current)) {
          break;
        }
        nodeSet.add(current);
        current = current.parent;
      }
    }
  });
  return nodeSet;
}

/**
 * 从所有的节点中找到筛选节点的所有父级和子级，用于展示
 * @param rootTreeNodes 所有的节点列表
 * @param filterTreeNode 筛选节点函数
 * @returns
 */
export function getTreeNodesWithAncestorsAndAllChildren(
  rootTreeNodes: DoublyTreeNode[],
  filterTreeNode: (node: DoublyTreeNode) => boolean,
) {
  const nodeSet = new Set<DoublyTreeNode>();
  // 使用广度优先，保证父节点先被遍历到
  const queue = new Queue(rootTreeNodes);
  for (;;) {
    const node = queue.dequeue();
    if (!node) {
      break;
    }
    // 如果命中则所有的祖先节点、子节点全部加入
    if (filterTreeNode(node)) {
      // 祖先节点和自己
      for (let current: DoublyTreeNode | null = node; ; ) {
        // 如果到头了就不跑了，如果在 nodeSet 中出现过说明他的父级也都已经在 nodeSet 中了，也不跑了
        if (!current || nodeSet.has(current)) {
          break;
        }
        nodeSet.add(current);
        current = current.parent;
      }
      // 子节点
      traverseDoublyTreeNode(node, child => {
        nodeSet.add(child);
      });
      continue;
    }
    node.children?.forEach(n => {
      queue.enqueue(n);
    });
  }
  return nodeSet;
}
