// @universe-design/react/es/color-picker/pickers/ColorPicker.js
import React, { useState, useMemo, useRef } from 'react';
import clsx from 'clsx';
import { ColorPickerStep } from '../type';
import { judgeEmptyColorGroup } from '../util';
import RightBoldOutlined from '@universe-design/icons-react/RightBoldOutlined';
import { useLocale } from '../../config-provider/hooks';
import { DEFAULT_COLOR_GROUP } from '../colors';
import { Palette } from './Palette';
import { MOREBTNICON } from '../const';
import { useComplementEvent } from '../../hooks/useComplementEvent';
import { ColorPickerProvider, useColorPicker, useColorPickerContext } from '../ColorPickerContext';
import { ColorPickerColorBlock } from '../components/ColorPickerColorBlock';
import '../style';

/**
 * @public
 * @docgen
 */

const Blocks = props => {
  const {
    colorGroup
  } = props;
  const {
    rootCssScope
  } = useColorPickerContext();
  const prefix = rootCssScope('-swatches');
  const isEmptyColorGroup = useMemo(() => judgeEmptyColorGroup(colorGroup), [colorGroup]);
  return /*#__PURE__*/React.createElement("div", {
    className: clsx(prefix, isEmptyColorGroup && `${prefix}-empty`)
  }, colorGroup.map((colors, index) => /*#__PURE__*/React.createElement("div", {
    key: index,
    className: `${prefix}-row`
  }, colors.map(color => /*#__PURE__*/React.createElement(ColorPickerColorBlock, {
    key: JSON.stringify(color),
    color: color
  })))));
};
export const ColorPicker = props => {
  const locale = useLocale();
  const {
    colorGroup = DEFAULT_COLOR_GROUP,
    showMoreColor = true,
    moreBtnText = locale.components.colorPicker.moreColor,
    className,
    style,
    onSaturationDragEnd,
    onHueDragEnd
  } = props;
  const [currentStep, setCurrentStep] = useState(ColorPickerStep.inHome);
  const colorPickerCtx = useColorPicker(props);
  const {
    rootCssScope
  } = colorPickerCtx;
  const prefix = rootCssScope();
  const isEmptyColorGroup = useMemo(() => judgeEmptyColorGroup(colorGroup), [colorGroup]);
  const containerRef = useRef(null);
  const moreColorBtnRef = useRef(null);
  useComplementEvent('pointerdown', {
    listener: React.useCallback(() => {
      setCurrentStep(ColorPickerStep.inHome);
    }, []),
    // 这里必须有两个 ref，是因为点击 moreColorBtn 的时候会跳转到另一页，在事件触发的过程中，moreColorBtn 已经消失了，所以会判断为点击了外面，这个写法是为了保证事件从 moreColor 出来也可以判断是在组件内部
    elements: React.useMemo(() => [containerRef, moreColorBtnRef.current], []),
    skipped: currentStep !== ColorPickerStep.inMoreColor
  });
  const renderMoreColorBtn = () => {
    if (showMoreColor === false) {
      return null;
    }
    return /*#__PURE__*/React.createElement("div", {
      ref: moreColorBtnRef,
      className: clsx(`${prefix}-more`, isEmptyColorGroup && `${prefix}-more-empty`),
      onClick: () => setCurrentStep(ColorPickerStep.inMoreColor)
    }, /*#__PURE__*/React.createElement("div", {
      className: `${prefix}-more-left`
    }, /*#__PURE__*/React.createElement("img", {
      src: MOREBTNICON
    }), /*#__PURE__*/React.createElement("span", null, moreBtnText)), /*#__PURE__*/React.createElement(RightBoldOutlined, null));
  };
  return /*#__PURE__*/React.createElement(ColorPickerProvider, {
    value: colorPickerCtx
  }, (() => {
    if (currentStep === ColorPickerStep.inHome) {
      return /*#__PURE__*/React.createElement("div", {
        ref: containerRef,
        className: clsx(prefix, className),
        style: style
      }, /*#__PURE__*/React.createElement(Blocks, {
        colorGroup: colorGroup
      }), renderMoreColorBtn());
    }
    return /*#__PURE__*/React.createElement(Palette, {
      onSaturationDragEnd: onSaturationDragEnd,
      onHueDragEnd: onHueDragEnd,
      ref: containerRef,
      className: className,
      style: style
    });
  })());
};
