// @universe-design/react/es/color-picker/pickers/SimpleBlocksColorPicker.js
import React from 'react';
import clsx from 'clsx';
import { udtoken_colorpicker_blue, udtoken_colorpicker_red, udtoken_colorpicker_orange, udtoken_colorpicker_indigo, udtoken_colorpicker_violet, udtoken_colorpicker_purple, udtoken_colorpicker_yellow, udtoken_colorpicker_green, udtoken_colorpicker_turquoise, udtoken_colorpicker_wathet, udtoken_colorpicker_neutral } from '@universe-design/styles/esm/universe-design/index.light.atom';
import { ColorPickerProvider, useColorPicker } from '../ColorPickerContext';
import { Space } from '../../space';
import { ColorPickerColorBlock } from '../components/ColorPickerColorBlock';
import '../style';

/**
 * @public
 * `<SimpleBlocksColorPicker />` 组件的属性。
 */

const defaults = {
  colorGroup: [[undefined, udtoken_colorpicker_red, udtoken_colorpicker_orange, udtoken_colorpicker_yellow, udtoken_colorpicker_green, udtoken_colorpicker_turquoise], [udtoken_colorpicker_blue, udtoken_colorpicker_wathet, udtoken_colorpicker_indigo, udtoken_colorpicker_purple, udtoken_colorpicker_violet, udtoken_colorpicker_neutral]]
};
export const SimpleBlocksColorPicker = props => {
  const {
    className,
    style,
    colorGroup = defaults.colorGroup
  } = props;
  const colorPickerCtx = useColorPicker(props);
  const {
    rootCssScope
  } = colorPickerCtx;
  return /*#__PURE__*/React.createElement(ColorPickerProvider, {
    value: colorPickerCtx
  }, /*#__PURE__*/React.createElement("div", {
    className: clsx(rootCssScope(), rootCssScope('-simple'), className),
    style: style
  }, /*#__PURE__*/React.createElement(Space, {
    direction: "vertical"
  }, colorGroup.map((row, i) => /*#__PURE__*/React.createElement(Space, {
    key: i,
    inline: false
  }, row.map(color => /*#__PURE__*/React.createElement(ColorPickerColorBlock, {
    key: typeof color === 'undefined' ? 'undefined' : JSON.stringify(color),
    color: color
  })))))));
};
