// @universe-design/styles/cjs/im/index.light.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.imtoken_message_bg_bubbles_blue = exports.imtoken_message_bg_audio_grey = exports.imtoken_message_bg_audio_blue = exports.imtoken_general_icon_encrypted_chat_normal = exports.imtoken_feed_topitem_hover_btn_press = exports.imtoken_feed_topitem_hover_btn_hover = exports.imtoken_feed_topitem_hover_btn_default = exports.imtoken_feed_topitem_default_btn_default = exports.imtoken_feed_topitem_bg_press = exports.imtoken_feed_topitem_bg_hover = exports.imtoken_feed_topitem_bg_default = exports.imtoken_feed_topitem_active_btn_press = exports.imtoken_feed_topitem_active_btn_hover = exports.imtoken_feed_topitem_active_btn_default = exports.imtoken_feed_text_pri_selected = exports.imtoken_feed_item_hover_btn_press = exports.imtoken_feed_item_hover_btn_hover = exports.imtoken_feed_item_hover_btn_default = exports.imtoken_feed_item_default_btn_default = exports.imtoken_feed_item_bg_press = exports.imtoken_feed_item_active_btn_press = exports.imtoken_feed_item_active_btn_hover = exports.imtoken_feed_item_active_btn_default = exports.imtoken_feed_icon_pri_selected = exports.imtoken_feed_fill_active = exports.imtoken_feed_bg_top = exports.imtoken_feed_bg_reaction = exports.imtoken_feed_bg_handle_later = exports.imtoken_feed_bg_delete = exports.imtoken_feed_bg_complete = exports.imtoken_feed_bg_body = exports.imtoken_dialog_bg_top_cricle = exports.imtoken_dialog_bg_top = exports.imtoken_bg_Aurora_Horizon = exports.imtoken_add_menu_icon_bg = exports.imtoken_AI_onboarding_Avatar_bg = exports.imtoken_AI_message_bg = exports.imtoken_AI_Profile_Avatar_bg = exports.im_btn_selected_text_red = exports.im_btn_selected_text_green = exports.im_btn_selected_line_gray = exports.im_btn_selected_bg_red_press = exports.im_btn_selected_bg_red_normal = exports.im_btn_selected_bg_red_hover = exports.im_btn_selected_bg_green_press = exports.im_btn_selected_bg_green_normal = exports.im_btn_selected_bg_green_hover = exports.im_btn_selected_bg_gray_press = exports.im_btn_selected_bg_gray_normal = exports.im_btn_selected_bg_gray_hover = void 0;
exports.imtoken_set_colour_Aurora_Lake = exports.imtoken_set_colour_Aurora_Horizon = exports.imtoken_set_colour_Aurora_Air = exports.imtoken_secure_badge_bg_normal = exports.imtoken_secret_chat_bg_topbar = exports.imtoken_red_envelope_card_text_title_yellow = exports.imtoken_red_envelope_card_text_placeholder_yellow = exports.imtoken_red_envelope_button_text_yellow = exports.imtoken_profile_text_normal = exports.imtoken_profile_icon_normal = exports.imtoken_profile_bg_press = exports.imtoken_profile_bg_normal = exports.imtoken_profile_bg_hover = exports.imtoken_message_voice_text_time_gray = exports.imtoken_message_voice_text_time_blue = exports.imtoken_message_voice_line_schedule_grey = exports.imtoken_message_voice_line_schedule_blue = exports.imtoken_message_text_pin = exports.imtoken_message_text_bubbles_grey = exports.imtoken_message_text_bubbles_blue = exports.imtoken_message_sidebar_group_bg_simple = exports.imtoken_message_sidebar_group_bg = exports.imtoken_message_pin_text_bg_pressed = exports.imtoken_message_pin_text_bg_hover = exports.imtoken_message_pin_text_bg_focus = exports.imtoken_message_pin_popover_line = exports.imtoken_message_pin_popover_bg = exports.imtoken_message_pin_list_popover_caret = exports.imtoken_message_icon_voice_grey_bg = exports.imtoken_message_icon_voice_grey = exports.imtoken_message_icon_voice_blue_bg = exports.imtoken_message_icon_voice_blue = exports.imtoken_message_icon_unread = exports.imtoken_message_icon_read = exports.imtoken_message_icon_pin_list = exports.imtoken_message_icon_pin = exports.imtoken_message_icon_lan = exports.imtoken_message_card_border = exports.imtoken_message_card_bg_body_embed = exports.imtoken_message_card_bg_body = exports.imtoken_message_bg_selected = exports.imtoken_message_bg_reaction_grey_simple = exports.imtoken_message_bg_reaction_grey_body = exports.imtoken_message_bg_reaction_grey = exports.imtoken_message_bg_reaction_blue = exports.imtoken_message_bg_pin = exports.imtoken_message_bg_location = exports.imtoken_message_bg_chat_simple = exports.imtoken_message_bg_chat = exports.imtoken_message_bg_bubbles_grey = void 0;
exports.imtoken_topic_group_comment_file = exports.imtoken_text_tips_link_pressed = exports.imtoken_text_tips_link_normal = exports.imtoken_text_tips_link_hover = exports.imtoken_t_entrance_hover = exports.imtoken_t_entrance_default = exports.imtoken_side_unread_bg_hover = exports.imtoken_side_unread_bg_default = exports.imtoken_side_unread_bg_active = exports.imtoken_set_colour_Aurora_Sky = exports.imtoken_set_colour_Aurora_Obsidian = void 0;
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
exports.im_btn_selected_bg_gray_hover = 'rgba(31, 35, 41, 0.1)';
exports.im_btn_selected_bg_gray_normal = 'rgba(31, 35, 41, 0.06)';
exports.im_btn_selected_bg_gray_press = 'rgba(31, 35, 41, 0.14)';
exports.im_btn_selected_bg_green_hover = '#d0f5ce';
exports.im_btn_selected_bg_green_normal = '#e4fae1';
exports.im_btn_selected_bg_green_press = '#95e599';
exports.im_btn_selected_bg_red_hover = '#fee3e2';
exports.im_btn_selected_bg_red_normal = '#fef0f0';
exports.im_btn_selected_bg_red_press = '#fdc6c4';
exports.im_btn_selected_line_gray = '#646a73';
exports.im_btn_selected_text_green = '#1a7526';
exports.im_btn_selected_text_red = '#e22e28';
exports.imtoken_AI_Profile_Avatar_bg = 'linear-gradient(to top, #ffffff 32%, #e1d6f9 66%, #b780e0 100%)';
exports.imtoken_AI_message_bg = '#f4f5f8';
exports.imtoken_AI_onboarding_Avatar_bg = 'rgba(0, 0, 0, 0)';
exports.imtoken_add_menu_icon_bg = '#ffffff';
exports.imtoken_bg_Aurora_Horizon = '#ecedee';
exports.imtoken_dialog_bg_top = 'linear-gradient(180deg, #e4f2ff 0%, #ffffff 100%)';
exports.imtoken_dialog_bg_top_cricle = 'linear-gradient(180deg, rgba(211, 229, 255, 0.64) 0%, rgba(255, 255, 255, 0.8) 100%)';
exports.imtoken_feed_bg_body = '#ffffff';
exports.imtoken_feed_bg_complete = '#10a893';
exports.imtoken_feed_bg_delete = '#e22e28';
exports.imtoken_feed_bg_handle_later = '#fad355';
exports.imtoken_feed_bg_reaction = 'rgba(31, 35, 41, 0.1)';
exports.imtoken_feed_bg_top = '#646a73';
exports.imtoken_feed_fill_active = 'rgba(51, 133, 255, 0.12)';
exports.imtoken_feed_icon_pri_selected = '#1456f0';
exports.imtoken_feed_item_active_btn_default = 'rgba(31, 35, 41, 0.08)';
exports.imtoken_feed_item_active_btn_hover = 'rgba(31, 35, 41, 0.13)';
exports.imtoken_feed_item_active_btn_press = 'rgba(31, 35, 41, 0.18)';
exports.imtoken_feed_item_bg_press = 'rgba(31, 35, 41, 0.12)';
exports.imtoken_feed_item_default_btn_default = '#eff0f1';
exports.imtoken_feed_item_hover_btn_default = '#d8d9db';
exports.imtoken_feed_item_hover_btn_hover = '#d0d1d3';
exports.imtoken_feed_item_hover_btn_press = '#cacbcd';
exports.imtoken_feed_text_pri_selected = '#1456f0';
exports.imtoken_feed_topitem_active_btn_default = '#ffffff';
exports.imtoken_feed_topitem_active_btn_hover = 'rgba(255, 255, 255, 0.65)';
exports.imtoken_feed_topitem_active_btn_press = 'rgba(31, 35, 41, 0.05)';
exports.imtoken_feed_topitem_bg_default = '#f2f3f5';
exports.imtoken_feed_topitem_bg_hover = '#e8e9eb';
exports.imtoken_feed_topitem_bg_press = '#e1e2e4';
exports.imtoken_feed_topitem_default_btn_default = '#ffffff';
exports.imtoken_feed_topitem_hover_btn_default = '#ffffff';
exports.imtoken_feed_topitem_hover_btn_hover = 'rgba(255, 255, 255, 0.6)';
exports.imtoken_feed_topitem_hover_btn_press = 'rgba(31, 35, 41, 0.05)';
exports.imtoken_general_icon_encrypted_chat_normal = '#373c43';
exports.imtoken_message_bg_audio_blue = 'rgba(20, 86, 240, 0.15)';
exports.imtoken_message_bg_audio_grey = 'rgba(31, 35, 41, 0.05)';
exports.imtoken_message_bg_bubbles_blue = '#d1e3ff';
exports.imtoken_message_bg_bubbles_grey = '#ebeced';
exports.imtoken_message_bg_chat = '#fcfcfc';
exports.imtoken_message_bg_chat_simple = '#fcfcfc';
exports.imtoken_message_bg_location = '#faeec4';
exports.imtoken_message_bg_pin = '#fcf5df';
exports.imtoken_message_bg_reaction_blue = 'rgba(20, 86, 240, 0.15)';
exports.imtoken_message_bg_reaction_grey = 'rgba(31, 35, 41, 0.05)';
exports.imtoken_message_bg_reaction_grey_body = 'rgba(31, 35, 41, 0.05)';
exports.imtoken_message_bg_reaction_grey_simple = 'rgba(31, 35, 41, 0.05)';
exports.imtoken_message_bg_selected = 'rgba(0, 0, 0, 0.02)';
exports.imtoken_message_card_bg_body = '#ffffff';
exports.imtoken_message_card_bg_body_embed = '#ffffff';
exports.imtoken_message_card_border = '#dee0e3';
exports.imtoken_message_icon_lan = '#ff811a';
exports.imtoken_message_icon_pin = '#10a893';
exports.imtoken_message_icon_pin_list = '#10a893';
exports.imtoken_message_icon_read = '#10a893';
exports.imtoken_message_icon_unread = '#8f959e';
exports.imtoken_message_icon_voice_blue = '#000000';
exports.imtoken_message_icon_voice_blue_bg = '#0442d2';
exports.imtoken_message_icon_voice_grey = '#000000';
exports.imtoken_message_icon_voice_grey_bg = '#373c43';
exports.imtoken_message_pin_list_popover_caret = '#fff8f0';
exports.imtoken_message_pin_popover_bg = '#f0f4ff';
exports.imtoken_message_pin_popover_line = '#ffffff';
exports.imtoken_message_pin_text_bg_focus = 'rgba(15, 133, 117, 0.1)';
exports.imtoken_message_pin_text_bg_hover = 'rgba(15, 133, 117, 0.1)';
exports.imtoken_message_pin_text_bg_pressed = 'rgba(15, 133, 117, 0.2)';
exports.imtoken_message_sidebar_group_bg = '#f5f6f7';
exports.imtoken_message_sidebar_group_bg_simple = '#f5f6f7';
exports.imtoken_message_text_bubbles_blue = '#1456f0';
exports.imtoken_message_text_bubbles_grey = '#51565d';
exports.imtoken_message_text_pin = '#10a893';
exports.imtoken_message_voice_line_schedule_blue = '#0442d2';
exports.imtoken_message_voice_line_schedule_grey = '#373c43';
exports.imtoken_message_voice_text_time_blue = '#0442d2';
exports.imtoken_message_voice_text_time_gray = '#1f2329';
exports.imtoken_profile_bg_hover = '#eff0f1';
exports.imtoken_profile_bg_normal = '#f5f6f7';
exports.imtoken_profile_bg_press = '#eff0f1';
exports.imtoken_profile_icon_normal = '#373c43';
exports.imtoken_profile_text_normal = '#2b2f36';
exports.imtoken_red_envelope_button_text_yellow = '#faedc2';
exports.imtoken_red_envelope_card_text_placeholder_yellow = 'rgba(252, 223, 126, 0.8)';
exports.imtoken_red_envelope_card_text_title_yellow = '#fcdf7e';
exports.imtoken_secret_chat_bg_topbar = '#383b42';
exports.imtoken_secure_badge_bg_normal = '#373c43';
exports.imtoken_set_colour_Aurora_Air = 'linear-gradient(45deg, #e3ecf7 0%, #d5dff6 100%)';
exports.imtoken_set_colour_Aurora_Horizon = '#ecedee';
exports.imtoken_set_colour_Aurora_Lake = 'linear-gradient(45deg, #476dae 0%, #4977ab 100%)';
exports.imtoken_set_colour_Aurora_Obsidian = 'linear-gradient(45deg, #3f4b64 0%, #697c96 100%)';
exports.imtoken_set_colour_Aurora_Sky = '#e6f1f4';
exports.imtoken_side_unread_bg_active = '#dee8f8';
exports.imtoken_side_unread_bg_default = '#f5f6f7';
exports.imtoken_side_unread_bg_hover = '#e4e5e7';
exports.imtoken_t_entrance_default = 'rgba(255, 255, 255, 0.7)';
exports.imtoken_t_entrance_hover = 'rgba(255, 255, 255, 0.9)';
exports.imtoken_text_tips_link_hover = '#336df4';
exports.imtoken_text_tips_link_normal = '#5083fb';
exports.imtoken_text_tips_link_pressed = '#94b4ff';
exports.imtoken_topic_group_comment_file = '#ffffff';
