// @byted-service/logger/lib/consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultStackCacheSize = exports.levels = exports.noop = exports.defaultEnableSecMark = exports.defaultFlushInterval = exports.defaultMetricsLevel = exports.defaultLogLevel = exports.defaultLevel = exports.version = void 0;
const env_1 = require("@byted-service/env");
// tslint:disable-next-line:no-var-requires
exports.version = require('../package.json').version;
exports.defaultLevel = 'info';
/**
 * logger默认的log level，低于这个level的数据不会生效
 * 取自: https://code.byted.org/gopkg/logs/blob/master/logger.go#L150
 */
exports.defaultLogLevel = 'debug';
exports.defaultMetricsLevel = 'warn';
exports.defaultFlushInterval = 10 * 1000; // 10s
exports.defaultEnableSecMark = process.env.LOGGER_ENABLE_SEC_MARK === 'true'
    ? true
    : process.env.LOGGER_ENABLE_SEC_MARK === 'false'
        ? false
        : (0, env_1.getIDC)() === 'useast5' && (0, env_1.getRegion)() === 'US-TTP'
            ? true
            : false;
const noop = () => {
    //
};
exports.noop = noop;
exports.levels = {
    trace: 10,
    debug: 20,
    info: 30,
    warn: 40,
    error: 50,
    fatal: 60,
};
exports.defaultStackCacheSize = 1000;
