// @byted-service/logger/lib/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = void 0;
const env_1 = require("@byted-service/env");
const util_1 = require("util");
const core_logger_1 = require("./core_logger");
const helper_1 = require("./helper");
let defaultLogger = null;
const formatLogOptions = (msg, inputData, options) => {
    inputData.logId = inputData.logId || '-';
    inputData.tags = options.tags
        ? Object.assign(Object.assign({}, options.tags), inputData.tags) : inputData.tags;
    inputData.tagKeys = options.tagKeys
        ? inputData.tagKeys
            ? options.tagKeys.concat(inputData.tagKeys)
            : options.tagKeys
        : inputData.tagKeys;
    inputData.message = options.prefix + msg;
    inputData.version = 'v1(7)';
    inputData.ts = Date.now();
    inputData.datetime = (0, helper_1.YYYYMMDDHHmmssSSS)(inputData.ts);
    inputData.psm = options.psm;
    inputData.cluster = (0, env_1.getCluster)();
    inputData.idc = (0, env_1.getIDC)();
    inputData.env = (0, env_1.getEnv)();
    inputData.deployStage = (0, env_1.getStage)();
    if (options.enableSecMark) {
        inputData.message = `${
        // Marked Tags
        (0, helper_1.toKV)(inputData.tags, inputData.tagKeys, ' ', true)}${
        // Msg
        inputData.message}`;
        inputData.tags = { current_version: (0, env_1.getImageVersion)() };
        inputData.tagKeys = undefined;
    }
    return inputData;
};
const buildFormatter = () => (_, inputData) => {
    return `${
    // Level
    inputData.level.toUpperCase()} ${
    // Time
    inputData.datetime} ${
    // Version
    inputData.version} ${
    // Location
    inputData.location} ${
    // Host
    (0, env_1.getHostIP)()} ${
    // PSM
    inputData.psm} ${
    // LogID
    inputData.logId} ${
    // Cluster
    inputData.cluster} ${
    // Stage
    inputData.deployStage} ${
    // Idc
    inputData.idc} ${
    // Tags
    (0, helper_1.toKV)(inputData.tags, inputData.tagKeys, ' ', false)}${
    // Msg
    inputData.message}\n`;
};
function getLocation(skipLocation, level, frameIndex) {
    if (skipLocation && level !== 'error' && level !== 'fatal') {
        return '-:-';
    }
    return (0, helper_1.getStackFilePath)(frameIndex + 1);
}
class Logger extends core_logger_1.CoreLogger {
    /** @deprecated use new Logger() instead */
    static get defaultLogger() {
        return (defaultLogger =
            defaultLogger ||
                Logger.createLogger({
                    disableDirFallbackWarning: true
                }));
    }
    /** @deprecated use new Logger() instead */
    static createLogger(options) {
        return new Logger(options);
    }
    constructor(options = {}) {
        var _a;
        const { level, streams, psm, prefix, tags, tagKeys, callSiteFrameIndex, enableSecMark } = (0, helper_1.formatOptions)(options);
        super(Object.assign(Object.assign({}, options), { level,
            streams, formatter: buildFormatter() }));
        this.psm = psm;
        this.prefix = prefix;
        this.tags = tags;
        this.tagKeys = tagKeys;
        this.callSiteFrameIndex = callSiteFrameIndex;
        this.enableSecMark = enableSecMark;
        this.skipLocation =
            (_a = options.skipLocation) !== null && _a !== void 0 ? _a : (!!process.env.LOGGER_SKIP_LOCATION || false);
        this.hooks = [];
    }
    /**
     * 获取同步 logger，可以用于无法异步写日志的情况。例如：
     * - process.on('uncaughtException')
     * - process.on('unhandledRejection')
     *
     * 请不要直接使用 console.log，因为 console.log 是否同步，根据环境不同而不同，具体见文档：https://nodejs.org/dist/latest-v12.x/docs/api/process.html#process_a_note_on_process_i_o
     *
     * - 方法出处：https://github.com/pinojs/pino/blob/master/docs/api.md#pinofinallogger-handler--function--finallogger
     * - 仅适用于 file log 写入，不支持 logagent
     */
    getFinalLogger() {
        return new Proxy(this, {
            get: (logger, key) => {
                if (key in logger.levels) {
                    return (message, ...splat) => {
                        logger.log({
                            level: key,
                            message,
                            splat,
                            location: getLocation(this.skipLocation, key, this.callSiteFrameIndex + 1)
                        });
                        logger.flushSync();
                    };
                }
                else if (key === 'log') {
                    return (args = {}) => {
                        var _a;
                        args.location =
                            args.location ||
                                getLocation(this.skipLocation, (_a = args.level) !== null && _a !== void 0 ? _a : this.defaultLevel, this.callSiteFrameIndex + 1);
                        logger.log(args);
                        logger.flushSync();
                    };
                }
                else {
                    return logger[key];
                }
            }
        });
    }
    log(args = {}) {
        const level = args.level || this.defaultLevel;
        if (!this.isLevelEnabled(level)) {
            return;
        }
        const message = args.message === undefined ? '' : args.message;
        const splat = args.splat || [];
        args.location = this.enableSecMark
            ? getLocation(this.skipLocation, level, this.callSiteFrameIndex + 1)
            : args.location ||
                getLocation(this.skipLocation, level, this.callSiteFrameIndex + 1);
        this.write(level, message, splat, args);
    }
    clone(options = {}) {
        const logger = super.clone(options);
        Object.setPrototypeOf(logger, this);
        logger.psm = options.psm || this.psm;
        logger.prefix = options.prefix || this.prefix;
        logger.tags = this.tags
            ? Object.assign(Object.assign({}, this.tags), options.tags) : options.tags;
        logger.tagKeys = this.tagKeys
            ? this.tagKeys.concat(options.tagKeys || [])
            : options.tagKeys;
        logger.callSiteFrameIndex =
            options.callSiteFrameIndex || this.callSiteFrameIndex;
        logger.setFormatter(buildFormatter());
        return logger;
    }
    addHook(hook) {
        if (typeof hook !== 'function') {
            throw new Error('hook must be a function');
        }
        this.hooks.push(hook);
    }
    write(level, message, splat, options = {}) {
        options.level = options.level || level;
        options.splat = options.splat || splat;
        const opts = formatLogOptions((0, util_1.format)(message, ...splat), options, {
            psm: this.psm,
            prefix: this.prefix,
            tags: this.tags,
            tagKeys: this.tagKeys,
            enableSecMark: this.enableSecMark
        });
        for (const hook of this.hooks) {
            hook(opts);
        }
        const msg = this.getFormatter().format(opts.message, opts);
        this.transport(level, msg, opts);
    }
    genLog(level) {
        if (level === 'fatal') {
            return (message, ...splat) => {
                this.write(level, message, splat, {
                    location: getLocation(this.skipLocation, level, this.callSiteFrameIndex + 2)
                });
            };
        }
        else {
            return (message, ...splat) => {
                this.write(level, message, splat, {
                    location: getLocation(this.skipLocation, level, this.callSiteFrameIndex + 1)
                });
            };
        }
    }
}
exports.Logger = Logger;
/** @deprecated */
Logger.getStackFilePath = helper_1.getStackFilePath;
