// @byted-nodex/metrics/lib/writer/pool.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WriterPool = void 0;
const generic_pool_1 = __importDefault(require("generic-pool"));
const util_1 = require("../util");
const defaultPoolingConfig = {
    max: 10,
    min: 1,
    acquireTimeoutMillis: 10000,
    idleTimeoutMillis: 60000,
};
class WriterPool {
    constructor(options) {
        this.logger = options.logger;
        this.options = { ...defaultPoolingConfig, ...options };
        this.pool = generic_pool_1.default.createPool({
            create: this._create.bind(this),
            destroy: this._destroy.bind(this),
        }, this.options);
    }
    async _create() {
        return new this.options.writer(this, this.options.writerOptions);
    }
    async _destroy(w) {
        await w.close();
    }
    async destroy(w) {
        return this.pool.destroy(w).catch(util_1.noop);
    }
    async acquire() {
        let w;
        for (let i = 0; i < this.options.max + 1; i++) {
            w = await this.pool.acquire();
            if (!w.closed) {
                break;
            }
            // writer 已经关闭了，释放掉
            await this.pool.destroy(w);
            w = undefined;
        }
        if (!w) {
            throw new Error('can not acquire writer');
        }
        if (!w.ready) {
            try {
                await w.init();
            }
            catch (e) {
                // 初始化失败也得释放掉资源
                await this.pool.destroy(w);
                throw e;
            }
        }
        return w;
    }
    async close() {
        await this.pool.drain();
        return this.pool.clear();
    }
    async release(w) {
        return this.pool.release(w);
    }
}
exports.WriterPool = WriterPool;
