// @byted-nodex/metrics/lib/writer/seqpacket.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SeqpacketWriter = void 0;
const util_1 = require("../util");
const defaultSocketWriteTimeout = 30 * 1000; // 写入体积很大，保不准会不会超时，默认大一些
const defaultSocketConnectTimeout = 5 * 1000;
/**
 * SeqpacketSender keeps connected SeqpacketSockets.
 */
class SeqpacketWriter {
    constructor(pool, options) {
        this.socket = null;
        this.ready = false;
        this.closed = false;
        this.logger = pool.logger;
        this.socketPath = options.socketPath;
        // node-unix-socket 是 rust 包，有用户本地环境 node 版本比较复杂，容易出现启动失败（ 缺对应架构的包 ）
        // 所以改成动态，尽可能减少影响
        this.Socket = require('node-unix-socket').SeqpacketSocket;
    }
    async close() {
        var _a;
        this.closed = true;
        this.ready = false;
        (_a = this.socket) === null || _a === void 0 ? void 0 : _a.destroy();
        this.socket = null;
    }
    async init() {
        await this.connect();
    }
    async write(buf) {
        if (!this.ready || !this.socket) {
            throw new Error('seqpacket socket is not ready or connection lost');
        }
        if (this.closed) {
            throw new Error('seqpacket connection has been closed');
        }
        await (0, util_1.timeoutPromise)((resolve, _) => {
            var _a;
            (_a = this.socket) === null || _a === void 0 ? void 0 : _a.write(buf, 0, buf.length, () => resolve(true));
        }, defaultSocketWriteTimeout, 'write data to socket timeout');
    }
    async connect() {
        if (this.socket || this.closed) {
            return;
        }
        const socket = (this.socket = new this.Socket());
        socket.unref();
        this.socket
            .on('error', e => {
            this.logger.error('seqpacket socket error', e);
        })
            .on('close', () => {
            this.logger.info('seqpacket socket has been closed');
            this.close();
        });
        try {
            await (0, util_1.timeoutPromise)((resolve, _) => {
                socket.connect(this.socketPath, () => resolve());
            }, defaultSocketConnectTimeout, 'seqpacket socket connect timeout');
            this.ready = true;
        }
        catch (err) {
            this.logger.error('Connect to %s failed with error: %s', this.socketPath, err);
            this.close();
            throw err;
        }
    }
}
exports.SeqpacketWriter = SeqpacketWriter;
