// @universe-design/react/src/image-viewer/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Basic from '../demo/0.basic';
import C1MaxMinScale from '../demo/1.max_min_scale';
import C2Plugin from '../demo/2.plugin';
import C3Error from '../demo/3.error';
import C4CustomToolbar from '../demo/4.custom-toolbar';
if (!(C0Basic as Story).storyName) {
  (C0Basic as Story).storyName = '0. 基础用法';
  (C0Basic as Story).parameters = {
    docs: {
      description: {
        story: '基础用法：简单的用法，只需要给 images 属性就行\n\n',
      },
    },
  };
}
if (!(C1MaxMinScale as Story).storyName) {
  (C1MaxMinScale as Story).storyName = '1. 图片缩放';
  (C1MaxMinScale as Story).parameters = {
    docs: {
      description: {
        story: '通过 `maxScale` 和 `minScale` 设置缩放的最大最小比例。\n\n',
      },
    },
  };
}
if (!(C2Plugin as Story).storyName) {
  (C2Plugin as Story).storyName = '2. 插件用法';
  (C2Plugin as Story).parameters = {
    docs: {
      description: {
        story: '插件用法：可以添加额外的插件来自行配置。\n\n',
      },
    },
  };
}
if (!(C3Error as Story).storyName) {
  (C3Error as Story).storyName = '3. errorFallback用法';
  (C3Error as Story).parameters = {
    docs: {
      description: {
        story: 'errorFallback 用法：设置 errorFallback 替换原本的错误显示内容。\n\n',
      },
    },
  };
}
if (!(C4CustomToolbar as Story).storyName) {
  (C4CustomToolbar as Story).storyName = '4. 自定义 ToolBar';
  (C4CustomToolbar as Story).parameters = {
    docs: {
      description: {
        story: '通过 slot 的方式实现完全自定义的 ToolBar。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/ImageViewer/官网示例',
  tags: ['autodocs'],
};

export { C0Basic, C1MaxMinScale, C2Plugin, C3Error, C4CustomToolbar };
