// @universe-design/react/src/date-picker/hooks/useFormatAndParser.ts
/**
 * 将 input 的输入值转换为 date
 * 将 value 转换为 string 提供给 input
 */
import isFunction from 'lodash/isFunction';
import { useLocale } from '../../config-provider/hooks';
import type { IFormat, IParse, MODE } from '../interface';
import moment from 'moment';

type PresetFormats = {
  year: string;
  month: string;
  date: string;
};

function getDefaultFormat(options: { mode?: MODE; showTime?: boolean; presetFormats: PresetFormats }) {
  const { mode, showTime = false, presetFormats } = options;
  switch (mode) {
    case 'year':
      return presetFormats.year;
    case 'month':
      return presetFormats.month;
    case 'date':
    default:
      return showTime ? `${presetFormats.date} HH:mm:ss` : presetFormats.date;
  }
}

export default ({
  format: formatProp,
  parse,
  mode,
  showTime,
}: {
  format?: IFormat;
  parse?: IParse;
  mode: MODE;
  showTime: boolean;
}) => {
  const {
    locale: localeName,
    components: { datePicker: locale },
  } = useLocale();
  const presetFormats = {
    year: 'YYYY',
    month: locale.longMonthFormat,
    date: locale.longDateFormat,
  };

  const defaultFormat = getDefaultFormat({
    mode,
    showTime,
    presetFormats,
  });
  const format = formatProp ?? defaultFormat;

  function formatValue(value?: moment.Moment) {
    if (!value) {
      return '';
    }

    if (isFunction(format)) {
      return format(value);
    }
    return value.clone().locale(localeName).format(format);
  }

  function parseValue(value?: string) {
    if (!value) {
      return null;
    }

    if (isFunction(parse)) {
      return parse(value);
    }
    /**
     * 对于 parser 而言，如果 format 类型为 (value: moment) => string
     * 此时是无法处理的 （设计上的一个漏洞）
     * 此时回退到 defaultFormat
     */
    const formatForParser = typeof format === 'string' ? format : defaultFormat;

    return moment(value, formatForParser, localeName, true);
  }

  return {
    formatValue,
    parseValue,
    format,
  };
};
