// @universe-design/react/src/input/Group.tsx
import * as React from 'react';
import clsx from 'clsx';
import type { DefaultSizeType } from '../util';
import { getDefaultSizeAbbr } from '../util';
import { useCssScope } from '../config-provider/hooks';
import { InputGroupSizeContextProvider } from '../config-provider/size';

import './style';

/**
 * 组合输入框： `Input.Group`
 */
export interface GroupProps {
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 自定义类名
   */
  className?: string;
  /**
   * `Input.Group`中所有的`Input`的大小
   * @defaultValue `medium`
   */
  size?: DefaultSizeType;
  /**
   * 是否用紧凑模式
   * @defaultValue false
   */
  compact?: boolean;
  /**
   * 子组件
   */
  children?: React.ReactNode;
  /**
   * 鼠标移入事件
   */
  onMouseEnter?: React.MouseEventHandler<HTMLSpanElement>;
  /**
   * 鼠标移出事件
   */
  onMouseLeave?: React.MouseEventHandler<HTMLSpanElement>;
  /**
   * 聚焦事件
   */
  onFocus?: React.FocusEventHandler<HTMLSpanElement>;
  /**
   * 失焦事件
   */
  onBlur?: React.FocusEventHandler<HTMLSpanElement>;
}

export const Group: React.FC<GroupProps> = props => {
  const { style, className, size = 'medium', compact, children, onMouseEnter, onMouseLeave, onFocus, onBlur } = props;
  const cssScope = useCssScope('input-group');
  const usedSize = getDefaultSizeAbbr(size, 'medium');
  const cls = clsx(
    cssScope(),
    {
      [cssScope(`--size-${usedSize}`)]: usedSize,
      [cssScope('--compact')]: compact,
    },
    className,
  );

  return (
    <InputGroupSizeContextProvider size={size}>
      <span
        className={cls}
        style={style}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
        onFocus={onFocus}
        onBlur={onBlur}
      >
        {children}
      </span>
    </InputGroupSizeContextProvider>
  );
};

export default Group;
