// @universe-design/react/src/input/Input.tsx
/** @jsxRuntime classic */
/** @jsx jsx */
/** @jsxFrag React.Fragment */
import React, { forwardRef, useRef, useState, useImperativeHandle, useCallback, useEffect } from 'react';
import clsx from 'clsx';
import EmptyFilled from '@universe-design/icons-react/EmptyFilled';
import stylex from '@universe-design/stylex';
import { sxOverflowEllipsis } from '../stylex';
import type { DefaultSizeType } from '../util/size';
import { getDefaultSizeAbbr } from '../util/size';
import { isFunction } from '../util/assert';
import type { StylexProps } from '../util';
import { focus } from '../util';
import { useComponentSize, useInputGroupSize, useCssScope } from '../config-provider/hooks';
import NativeInput from '../shared/components/Input';
import CounterComp from './components/Counter';
import { useInput } from '@universe-design/react-input';
import type { ChangeHandler, InputCounterProps } from '@universe-design/react-input';
import type { TagProps } from '../tag';
import { FadeTransition } from '../shared/components/transition/FadeTransition';
import { useMouseOver } from '../hooks';
import { resolveOnChange } from './utils';
import { eventContains, isOverflowed } from '../util/dom';
import { useTextOverflowPopperSidecar } from '../config-provider/hooks/useTextOverflowPopperSidecar';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx } from '@universe-design/stylex/react';

import './style';

export interface InputProps
  extends Omit<React.InputHTMLAttributes<HTMLInputElement>, 'size' | 'prefix'>,
    InputCounterProps,
    StylexProps {
  /**
   * 类名前缀
   */
  prefixCls?: string;
  /**
   * 控件大小
   */
  size?: DefaultSizeType;
  /**
   * 带标签的 input，设置前置标签
   */
  addonBefore?: React.ReactNode;
  /**
   * 带标签的 input，设置后置标签
   */
  addonAfter?: React.ReactNode;
  /**
   * 带有前缀图标的 input
   */
  prefix?: React.ReactNode;
  /**
   * 带有后缀图标的 input
   */
  suffix?: React.ReactNode;
  /**
   * 可以点击清除图标删除内容
   */
  allowClear?: boolean;
  /**
   * 是否错误状态
   * @defaultValue false
   */
  error?: boolean;
  /**
   * 是否禁用状态
   * @defaultValue false
   */
  disabled?: boolean;
  /**
   * 是否只读状态
   * @defaultValue false
   */
  readOnly?: boolean;
  /**
   * 输入框内容
   */
  value?: string;
  /**
   * 输入框默认内容
   */
  defaultValue?: string;
  /**
   * 声明 input 类型，同原生 input 标签的`type`属性，见：[MDN](https://developer.mozilla.org/zh-CN/docs/Web/HTML/Element/input#属性) (请直接使用`Input.TextArea`代替`type="textarea"`)
   * @defaultValue `text`
   */
  type?: string;
  /**
   * 输入框内容变化时的回调
   */
  onChange?: React.ChangeEventHandler<HTMLInputElement>;
  /**
   * 按下回车的回调
   */
  onPressEnter?: React.KeyboardEventHandler<HTMLInputElement>;
  /**
   * 点击清除按钮后的回调
   */
  onClear?: () => void;
}

export const Input = forwardRef<HTMLInputElement, InputProps>((props, ref) => {
  const {
    style,
    prefixCls,
    className,
    readOnly,
    placeholder,
    disabled,
    size,
    prefix,
    suffix,
    addonBefore,
    addonAfter,
    defaultValue = '',
    value: propsValue,
    allowClear,
    type,
    error,
    onChange,
    onKeyDown,
    onPressEnter,
    onClear,
    counter: counterProp,
    slicer,
    maxLength: maxLengthProp,
    showCount,
    truncateExcessWords,
    onCompositionStart: onCompositionStartProp,
    onCompositionEnd: onCompositionEndProp,
    sx,
    ...otherProps
  } = props;

  const cssScope = useCssScope('input');
  const defaultSize = useComponentSize();
  const inputGroupSize = useInputGroupSize();
  const [focused, setFocused] = useState<boolean>(false);
  const inputRef = useRef<HTMLInputElement>(null);
  const timeout = useRef<ReturnType<typeof setTimeout>>();

  const { isMouseOver, handleMouseEnter, handleMouseLeave } = useMouseOver();

  const currentSize = getDefaultSizeAbbr(size || inputGroupSize || defaultSize);

  const tagSize: TagProps['size'] = currentSize === 'lg' ? 'medium' : 'small';

  const handleInputChange: ChangeHandler<HTMLInputElement> = useCallback(
    (ref, event, value) => resolveOnChange(ref, event, onChange, value),
    [onChange],
  );

  const { nativeMaxLength, value, setValue, count, onCompositionEnd, onCompositionStart, internalOnChange } =
    useInput<HTMLInputElement>({
      maxLength: maxLengthProp,
      showCount,
      truncateExcessWords,
      counter: counterProp,
      slicer,
      defaultValue,
      inputRef,
      onChange: handleInputChange,
      onCompositionStart: onCompositionStartProp,
      onCompositionEnd: onCompositionEndProp,
      focused,
      ...('value' in props ? { value: propsValue } : {}),
    });

  const placeholderRef = React.useRef<HTMLDivElement>(null);

  const stop = useCallback(() => {
    timeout.current && clearTimeout(timeout.current);
  }, []);

  const start = useCallback(() => {
    timeout.current && clearTimeout(timeout.current);
    timeout.current = setTimeout(() => {
      if (
        inputRef.current &&
        inputRef.current.getAttribute('type') === 'password' &&
        inputRef.current.getAttribute('value')
      ) {
        inputRef.current.removeAttribute('value');
      }
      // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    }, 13);
  }, []);

  useEffect(() => {
    stop();
    start();
    return () => stop();
  }, [value, start, stop, type]);

  useImperativeHandle<HTMLInputElement | null, HTMLInputElement | null>(ref, () => inputRef.current);

  const renderAddonBefore = (cls: string) => {
    if (addonBefore) {
      return <span className={`${cls}__addon-before`}>{addonBefore}</span>;
    }
    return null;
  };

  const renderAddonAfter = (cls: string) => {
    if (addonAfter) {
      return <span className={`${cls}__addon-after`}>{addonAfter}</span>;
    }
    return null;
  };

  const renderPrefix = (cls: string) => {
    if (prefix) {
      return <span className={`${cls}__prefix`}>{prefix}</span>;
    }
    return null;
  };

  const renderSuffix = (cls: string) => {
    if (suffix || showCount) {
      return (
        <span className={`${cls}__suffix`}>
          {suffix}
          <CounterComp
            // marginLeft 会根据 input Size 变化吗
            style={{ marginLeft: suffix ? 8 : 0 }}
            showCount={showCount}
            count={count}
            maxLength={maxLengthProp}
            tagSize={tagSize}
          />
        </span>
      );
    }
    return null;
  };

  const onFocusChange = useCallback(
    isFocus => {
      stop();
      start();
      setFocused(isFocus);
    },
    [start, stop],
  );

  const handleClear = (e: React.MouseEvent<HTMLSpanElement, MouseEvent>) => {
    if (!('value' in props)) {
      setValue('');
    }
    if (inputRef.current) {
      resolveOnChange(inputRef.current, e, onChange, '');
    }
    inputRef.current?.focus();
    isFunction(onClear) && onClear();
  };

  const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    /* istanbul ignore else  */
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    if (e.keyCode === 13 && onPressEnter) {
      onPressEnter(e);
    }
    /* istanbul ignore else  */
    if (onKeyDown) {
      onKeyDown(e);
    }
  };

  const renderClearIcon = (cls: string) => {
    if (allowClear && !disabled) {
      // 输入框激活态非 hover 状态不展示 query 清除按钮
      const isShow = isMouseOver && Boolean(value);
      const clx = clsx({
        [`${cls}__clear-icon`]: true,
        [`${cls}__clear-icon-show`]: isShow,
      });
      return (
        <FadeTransition in={isShow} timeout={400}>
          <span className={clx} onClick={handleClear}>
            <EmptyFilled />
          </span>
        </FadeTransition>
      );
    }
    return null;
  };

  const TextOverflowPopper = useTextOverflowPopperSidecar();
  // 当value 为空时才应该显示 tooltip
  const isPlaceholderVisible = Boolean(!value && placeholder);
  const [isPlaceholderTooltipVisible, setIsPlaceholderTooltipVisible] = React.useState(false);

  const renderInputWrapper = () => {
    const cls = prefixCls || cssScope();
    const rootCls = clsx(
      {
        [cls]: true,
        [`${cls}--error`]: error,
        [`${cls}-addon`]: addonAfter || addonBefore,
        [`${cls}--focused`]: focused,
        [`${cls}--disabled`]: disabled,
        [`${cls}--size-${currentSize}`]: currentSize,
      },
      className,
      stylex(stylex.create({ width: '100%' }), ...(sx ?? [])),
    );

    return (
      <div style={style} className={rootCls}>
        {renderAddonBefore(cls)}
        <label
          // 用 label 是为了嵌套 label 情况下能正常使用
          className={`${cls}-input-wrap`}
          onMouseEnter={handleMouseEnter}
          onMouseLeave={handleMouseLeave}
          onMouseDown={ev => {
            if (ev.target === inputRef.current) {
              // 如果是点击 input 的话就不需要做任何操作
              return;
            }
            // 如果事件是 wrap 里面的元素
            if (eventContains(ev.currentTarget, ev.nativeEvent)) {
              // 点击 wrap 内部元素不会让 input 失焦（mousedown 的 blur 逻辑会在这段同步代码之后执行）
              ev.preventDefault();
              // 点击 wrap 内部元素 focus 到 input 上
              focus(inputRef, { selectTextIfInput: false });
            }
          }}
          onClick={ev => {
            // 如果是 wrap 内部元素触发的 onClick，我们禁用 label 默认行为，因为 input 会触发 onClick，避免 bubble 两次（有对 prefix 和 suffix 误伤的概率）。
            if (eventContains(ev.currentTarget, ev.nativeEvent)) {
              ev.preventDefault();
            }
          }}
        >
          {renderPrefix(cls)}
          <TextOverflowPopper
            title={<>{placeholder}</>}
            disabled={disabled}
            checkOverflow={() => {
              if (placeholderRef.current) {
                return isOverflowed(placeholderRef.current);
              }
              return false;
            }}
            visible={isPlaceholderTooltipVisible && isPlaceholderVisible}
            onVisibleChange={visible => {
              setIsPlaceholderTooltipVisible(visible);
            }}
          >
            <div
              className={clsx(
                `${cls}-input__placeholder__wrapper`,
                stylex(
                  stylex.create({
                    position: 'relative',
                    flexGrow: 1,
                    minWidth: 0,
                    display: 'flex',
                    alignItems: 'center',
                  }),
                ),
              )}
              onMouseLeave={() => {
                /**
                 * 在 mouseleave 时，可能不会触发 tooltip 的 visibleChange
                 * 所以这里必须手动重置一下
                 */
                setIsPlaceholderTooltipVisible(false);
              }}
            >
              <div
                ref={placeholderRef}
                className={clsx(
                  `${cls}-input__placeholder`,
                  stylex(
                    sxOverflowEllipsis,
                    stylex.create({
                      position: 'absolute',
                      left: 0,
                      top: 0,
                      bottom: 0,
                      maxWidth: '100%',
                      visibility: 'hidden',
                    }),
                  ),
                )}
              >
                {placeholder}
              </div>
              <NativeInput
                {...otherProps}
                type={type}
                ref={inputRef}
                readOnly={readOnly}
                disabled={disabled}
                value={value}
                onChange={internalOnChange}
                onKeyDown={handleKeyDown}
                onFocusChange={onFocusChange}
                maxLength={nativeMaxLength}
                onCompositionEnd={onCompositionEnd}
                onCompositionStart={onCompositionStart}
                placeholder={placeholder}
                sx={[
                  stylex.create({
                    width: '100%',
                    minWidth: 20,
                  }),
                ]}
              />
            </div>
          </TextOverflowPopper>
          {renderClearIcon(cls)}
          {renderSuffix(cls)}
        </label>
        {renderAddonAfter(cls)}
      </div>
    );
  };

  return renderInputWrapper();
});
if (__DEV__) {
  Input.displayName = 'Input';
}
