// @universe-design/react/src/input/Password.tsx
import React, { useState } from 'react';
import type { InputProps } from './Input';
import { Input } from './Input';
import VisibleOutlined from '@universe-design/icons-react/VisibleOutlined';
import VisibleLockOutlined from '@universe-design/icons-react/VisibleLockOutlined';
import { useCssScope } from '../config-provider/hooks';
import { omit } from '../util';
import { Button } from '../button';

import './style';

/**
 * 用于密码输入框 `Input.Password`
 */
export interface PasswordProps extends InputProps {
  /**
   * 自定义类名前缀
   */
  readonly inputPrefixCls?: string;
  /**
   * 查看密码 icon 的触发方式，支持 `click` 或 `hover`
   */
  readonly action?: string;
  /**
   * 是否显示切换按钮
   * @defaultValue true
   */
  visibilityToggle?: boolean;
  /**
   * 自定义切换按钮
   * @defaultValue `(visible) => (visible ? <VisibleOutlined /> : <VisibleLockOutlined />`
   */
  iconRender?: (visible: boolean) => React.ReactNode;
}

const ActionMap: Record<string, string> = {
  click: 'onClick',
  hover: 'onMouseOver',
};

export const Password = React.forwardRef<HTMLInputElement, PasswordProps>((props, ref) => {
  const {
    action = 'click',
    size,
    visibilityToggle = true,
    iconRender = (visible: boolean) =>
      visible ? (
        <Button type="icon" size="small" icon={<VisibleOutlined />} />
      ) : (
        <Button type="icon" size="small" icon={<VisibleLockOutlined />} />
      ),
    ...restProps
  } = props;
  const [visible, setVisible] = useState(false);
  const cssScope = useCssScope('input-password');
  const prefixCls = cssScope();

  const onVisibleChange = () => {
    const { disabled } = props;
    if (disabled) {
      return;
    }

    setVisible(!visible);
  };

  const preventDefault = (e: MouseEvent) => e.preventDefault();

  const renderIcon = () => {
    const iconTrigger = ActionMap[action];
    const icon = iconRender(visible);
    const iconProps = {
      [iconTrigger]: onVisibleChange,
      key: 'passwordIcon',
      onMouseDown: (e: MouseEvent) => preventDefault(e),
      onMouseUp: (e: MouseEvent) => preventDefault(e),
    };
    return React.cloneElement(React.isValidElement(icon) ? icon : <span>{icon}</span>, iconProps);
  };

  const suffixIcon = visibilityToggle && renderIcon();
  const omittedProps = {
    ...omit(restProps, ['suffix']),
    type: visible ? 'text' : 'password',
    prefixCls,
    suffix: suffixIcon,
  };
  return <Input ref={ref} {...omittedProps} size={size} />;
});
