// @universe-design/react/src/input/TextArea.tsx
import type { CSSProperties } from 'react';
import React, { useCallback, useImperativeHandle, useState } from 'react';
import clsx from 'clsx';
import stylex from '@universe-design/stylex';
import { cssTypographyBody0 } from '../stylex';
import EmptyFilled from '@universe-design/icons-react/EmptyFilled';
import { useIsomorphicLayoutEffect as useLayoutEffect } from '../hooks';
import { useResizeObserver } from '@universe-design/react-utilities';
import { useClassPrefix, useCssScope } from '../config-provider/hooks';
import { isFunction, isNonNullable, KeyCode } from '../util';
import CounterComp from './components/Counter';
import { useInput, useFocus, calculateTextAreaHeight } from '@universe-design/react-input';
import type { ChangeHandler, InputCounterProps } from '@universe-design/react-input';

import './style';
import { resolveOnChange } from './utils';

export type TextAreaAutoSizeConfig = { minRows?: number; maxRows?: number };

/**
 * @public
 */
export interface TextAreaProps
  extends Omit<React.TextareaHTMLAttributes<HTMLTextAreaElement>, 'size' | 'prefix' | 'maxLength'>,
    InputCounterProps {
  /**
   * 可以点击清除图标删除内容
   * @defaultValue false
   */
  allowClear?: boolean;
  /**
   * 自适应内容高度，可设置为`true`、`false` 、对象 `{ minRows: 2, maxRows: 6 }`
   * @defaultValue false
   */
  autoSize?: boolean | TextAreaAutoSizeConfig;
  /**
   * 输入框内容
   */
  value?: string;
  /**
   * 是否错误状态
   * @defaultValue false
   */
  error?: boolean;
  /**
   * 输入框默认内容
   */
  defaultValue?: string;
  /**
   * 按下回车的回调
   */
  onPressEnter?: (e: React.KeyboardEvent<HTMLTextAreaElement>) => void;
  /**
   * 点击清除按钮后的回调
   */
  onClear?: () => void;
  /**
   * 带有后缀图标的 textarea
   */
  suffix?: React.ReactNode;
  /**
   * 调整文本框尺寸的方式
   * @defaultValue 'vertical'
   */
  resize?: React.CSSProperties['resize'];
}

export const TextArea = React.forwardRef<HTMLTextAreaElement, TextAreaProps>((props, ref) => {
  const {
    allowClear,
    autoSize,
    showCount,
    maxLength: maxLengthProp,
    value: valueProp,
    defaultValue,
    style,
    error,
    className,
    onKeyDown,
    onPressEnter,
    onChange,
    onClear,
    suffix,
    truncateExcessWords = false,
    counter: counterProp,
    slicer,
    onCompositionEnd: onCompositionEndProp,
    onCompositionStart: onCompositionStartProp,
    autoFocus,
    resize,
    ...resetProps
  } = props;

  const rootPrefixClass = useClassPrefix();
  const cssScope = useCssScope('textarea');
  const [textareaResized, setTextareaResized] = useState<boolean>(false);
  const textareaRef = useResizeObserver<HTMLTextAreaElement>(
    entry => {
      /**
       * 监控 textarea 的 style 属性上是否有 height 属性
       * textarea 只有在拖拽改变大小之后才会有 style height 属性
       * 此时，更改父容器为 height: auto 以自适应高度
       */
      if ((entry.target as HTMLTextAreaElement).style.height) {
        setTextareaResized(true);
      }
    },
    {
      disabled: Boolean(autoSize),
    },
  );
  const [textAreaStyle, setTextAreaStyle] = useState<CSSProperties>({});
  const [focused, setFocused] = useState(false);

  const handleInputChange: ChangeHandler<HTMLTextAreaElement> = useCallback(
    (ref, event, value) => resolveOnChange(ref, event, onChange, value),
    [onChange],
  );

  const { nativeMaxLength, value, setValue, count, onCompositionEnd, onCompositionStart, internalOnChange, maxLength } =
    useInput<HTMLTextAreaElement>({
      maxLength: maxLengthProp,
      showCount,
      truncateExcessWords,
      counter: counterProp,
      slicer,
      defaultValue,
      inputRef: textareaRef,
      onChange: handleInputChange,
      onCompositionStart: onCompositionStartProp,
      onCompositionEnd: onCompositionEndProp,
      focused,
      ...('value' in props ? { value: valueProp } : {}),
    });
  useFocus({
    inputRef: textareaRef,
    onFocusChange: f => setFocused(f),
    autoFocus: Boolean(autoFocus),
  });

  useImperativeHandle<HTMLTextAreaElement | null, HTMLTextAreaElement | null>(ref, () => textareaRef.current);

  const adjustSize = useCallback(() => {
    if (autoSize && textareaRef.current) {
      const config: TextAreaAutoSizeConfig = {
        minRows: undefined,
        maxRows: undefined,
      };
      if (typeof autoSize === 'object') {
        config.minRows = autoSize.minRows;
        config.maxRows = autoSize.maxRows;
      }
      const currentStyle = calculateTextAreaHeight(textareaRef.current, false, config.minRows, config.maxRows);
      if (currentStyle) {
        setTextAreaStyle(currentStyle);
      }
    } else {
      setTextAreaStyle({});
    }
  }, [autoSize, textareaRef]);

  const resizeRef = useResizeObserver<HTMLDivElement>(adjustSize); // , {disabled: !autoSize})

  useLayoutEffect(adjustSize, [adjustSize, value]);

  const handleKeyDown = (e: React.KeyboardEvent<HTMLTextAreaElement>) => {
    /* istanbul ignore else  */
    if (e.keyCode === KeyCode.ENTER) {
      isFunction(onPressEnter) && onPressEnter(e);
    }
    isFunction(onKeyDown) && onKeyDown(e);
  };

  const handleClear = (e: React.MouseEvent<HTMLElement, MouseEvent>) => {
    /* istanbul ignore else  */
    if (!('value' in props)) {
      setValue('');
    }
    /* istanbul ignore else  */
    if (isFunction(onChange) && textareaRef.current) {
      const eventObj = Object.create(e);
      eventObj.target = textareaRef.current;
      eventObj.currentTarget = textareaRef.current;
      const originalValue = textareaRef.current.value;
      textareaRef.current.value = '';
      onChange(eventObj as React.ChangeEvent<HTMLTextAreaElement>);
      textareaRef.current.value = originalValue;
    }
    textareaRef.current?.focus();
    isFunction(onClear) && onClear();
  };

  const renderSuffix = () => (
    <span className={cssScope('-suffix')}>
      {suffix && <span className={cssScope('-suffix-slot')}>{suffix}</span>}
      <CounterComp showCount={Boolean(showCount)} count={count} disabled={props.disabled} maxLength={maxLengthProp} />
    </span>
  );

  const showClearIcon = allowClear && !props.disabled && value;

  const renderClearIcon = () => {
    if (showClearIcon) {
      return (
        <span className={cssScope('__clear-icon')} onClick={handleClear}>
          <EmptyFilled />
        </span>
      );
    }
    return null;
  };

  /**
   * 用户传入的 error true 优先级很高（可能存在输入值不符合某个 form 规则而不只是 maxLength 的情况）
   * count > maxLength error 的优先级也很高
   * 只要有其一就需要标红
   */
  const isError = error || (isNonNullable(maxLength) && count > maxLength);
  const cls = clsx(
    {
      [cssScope()]: true,
      [cssScope('--error')]: isError,
      [cssScope('--show-count')]: showCount,
      [cssScope('--show-clear')]: showClearIcon,
      [cssScope('--disabled')]: props.disabled,
    },
    className,
  );

  return (
    <div className={cls} style={{ ...style, ...(textareaResized ? { height: 'auto' } : {}) }} ref={resizeRef}>
      {renderClearIcon()}
      <textarea
        {...resetProps}
        style={{ ...textAreaStyle, resize }}
        ref={textareaRef}
        className={clsx(`${rootPrefixClass}-scrollbar`, stylex(cssTypographyBody0))}
        onChange={internalOnChange}
        onKeyDown={handleKeyDown}
        onCompositionEnd={onCompositionEnd}
        onCompositionStart={onCompositionStart}
        value={value}
        maxLength={nativeMaxLength}
      />
      {renderSuffix()}
    </div>
  );
});
