// @universe-design/styles/ts/parse-theme.ts
import { TinyColor } from '@universe-design/colors';

export enum TokenMetaType {
  Raw = 'Raw',
  MappingColor = 'MappingColor',
  LinearGradient = 'LinearGradient',
  ConicGradient = 'ConicGradient',
  RadialGradient = 'RadialGradient',
  Shadow = 'Shadow',
  Length = 'Length',
  FamilyName = 'FamilyName',
}

export interface TokenMetaRawColor {
  type: TokenMetaType.Raw;
  raw: string;
  desc?: string;
  designName?: string;
}

export interface TokenMetaMappingColor {
  type: TokenMetaType.MappingColor;
  from: string;
  alpha?: number;
  desc?: string;
  designName?: string;
}

export interface TokenMetaLinearGradient {
  type: TokenMetaType.LinearGradient;
  // 角度，默认是 0
  angle: string;
  stops: {
    color: TokenMetaMappingColor | TokenMetaRawColor;
    // 0 - 1
    percentage: number;
  }[];
  desc?: string;
  designName?: string;
}
export interface TokenMetaConicGradient {
  type: TokenMetaType.ConicGradient;
  // 角度，默认是 0
  angle: string;
  stops: {
    color: TokenMetaMappingColor | TokenMetaRawColor;
    // 0 - 1
    percentage: number;
  }[];
  desc?: string;
  designName?: string;
}

export interface TokenMetaRadialGradient {
  type: TokenMetaType.RadialGradient;
  angle: string;
  stops: Array<{
    color: TokenMetaMappingColor | TokenMetaRawColor;
    percentage: number;
  }>;
  desc?: string;
  designName?: string;
}

export interface TokenMetaShadow {
  type: TokenMetaType.Shadow;
  items: {
    color: TokenMetaMappingColor | TokenMetaRawColor;
    offset: [number, number];
    blur: number;
    spread: number;
    inset: boolean;
  }[];
  desc?: string;
  designName?: string;
}

export interface TokenMetaLength {
  type: TokenMetaType.Length;
  value: string;
  desc?: string;
  designName?: string;
}

export interface TokenMetaFamilyName {
  type: TokenMetaType.FamilyName;
  familyName: string;
  designName?: string;
}

export type TokenMeta =
  | string
  | TokenMetaRawColor
  | TokenMetaMappingColor
  | TokenMetaLinearGradient
  | TokenMetaConicGradient
  | TokenMetaRadialGradient
  | TokenMetaShadow
  | TokenMetaFamilyName
  | TokenMetaLength;

export function parseTheme<Theme extends Record<string, TokenMeta>>(theme: Theme): { [K in keyof Theme]: string } {
  const plainTheme: Record<string, string> = {};

  const getRealValue = (meta: TokenMeta, key?: string): string => {
    if (key) {
      const finalValue = plainTheme[key];
      if (finalValue !== undefined) {
        return finalValue;
      }
    }
    let ret: string;
    if (typeof meta === 'string') {
      ret = meta;
    } else if (meta.type === TokenMetaType.Raw) {
      ret = meta.raw;
    } else if (meta.type === TokenMetaType.MappingColor) {
      const color = new TinyColor(getRealValue(theme[meta.from], meta.from));
      if (meta.alpha !== undefined && meta.alpha !== 1) {
        ret = color.setAlpha(meta.alpha).toRgbString();
      } else {
        ret = color.toRgbString();
      }
    } else if (meta.type === TokenMetaType.LinearGradient) {
      ret = `linear-gradient(${meta.angle}, ${meta.stops
        .map(stop => `${getRealValue(stop.color)} ${stop.percentage * 100}%`)
        .join(', ')})`;
    } else if (meta.type === TokenMetaType.ConicGradient) {
      ret = `conic-gradient(${meta.angle}, ${meta.stops
        .map(stop => `${getRealValue(stop.color)} ${stop.percentage * 100}%`)
        .join(', ')})`;
    } else if (meta.type === TokenMetaType.RadialGradient) {
      ret = `radial-gradient(${meta.angle}, ${meta.stops
        .map(stop => `${getRealValue(stop.color)} ${stop.percentage * 100}%`)
        .join(', ')})`;
    } else if (meta.type === TokenMetaType.Length) {
      ret = meta.value;
    } else if (meta.type === TokenMetaType.FamilyName) {
      ret = meta.familyName.replace(/"/g, '\\"');
    } else {
      ret = meta.items
        .map(
          item => `${item.offset[0]}px ${item.offset[1]}px ${item.blur}px ${item.spread}px ${getRealValue(item.color)}`,
        )
        .join(', ');
    }
    if (key) {
      plainTheme[key] = ret;
    }
    return ret;
  };

  Object.entries(theme).forEach(([key, meta]) => {
    getRealValue(meta, key);
  });
  return plainTheme as { [K in keyof Theme]: string };
}
