// @universe-design/styles/ts/TokensContext.tsx
import * as React from 'react';
import allLight from './all.light';
import { parseTheme, TokenMeta } from './parse-theme';

export type TokensContextType = Partial<typeof allLight>;

const TokensContext = React.createContext<TokensContextType>({});

if (process.env.NODE_ENV !== 'production') {
  TokensContext.displayName = 'UDTokensContext';
}

// Simulate cssvar
export const TokensProvider: React.FC<{
  tokens: Partial<Record<keyof typeof allLight, TokenMeta>>;
}> = ({ tokens: tokensProp, children }) => {
  const outerTokens = React.useContext(TokensContext);
  const tokens = React.useMemo(
    () => parseTheme({ ...outerTokens, ...tokensProp }),
    [outerTokens, tokensProp]
  );
  return (
    <TokensContext.Provider value={tokens}>{children}</TokensContext.Provider>
  );
};

export function useTokens() {
  return React.useContext(TokensContext);
}

export type WithTokensProps<P> = Omit<P, 'tokens'> & {
  tokens: TokensContextType;
};

type PropsWithoutTokens<P> = 'tokens' extends keyof P
  ? Pick<P, Exclude<keyof P, 'tokens'>>
  : P;

// For class component
export function withTokens<
  CC extends React.ComponentClass<WithTokensProps<{}>>
>(
  C: CC
): React.ForwardRefExoticComponent<
PropsWithoutTokens<React.ComponentPropsWithoutRef<CC>> &
React.RefAttributes<InstanceType<CC>>
>;

// For functional component
export function withTokens<P>(C: React.FC<WithTokensProps<P>>): React.FC<P>;

// For forwardRef functional component
export function withTokens<TRef, P>(
  C: React.ForwardRefExoticComponent<
  React.PropsWithoutRef<WithTokensProps<P>> & React.RefAttributes<TRef>
  >
): React.ForwardRefExoticComponent<
React.PropsWithoutRef<P> & React.RefAttributes<TRef>
>;

export function withTokens<TRef, P>(C: React.FC<WithTokensProps<P>>) {
  const displayName = C.displayName || C.name;

  const WrapperC = React.forwardRef<TRef, P>((props, ref) => (
    <TokensContext.Consumer>
      {tokens => <C {...props} tokens={tokens} ref={ref} />}
    </TokensContext.Consumer>
  ));

  WrapperC.displayName = `withTokens(${displayName})`;

  return WrapperC;
}
