// @universe-design/react/src/popover/index.tsx
import type { ReactNode } from 'react';
import React, { forwardRef, useMemo } from 'react';
import clsx from 'clsx';
import stylex from '@universe-design/stylex';
import { cssFlexBase, cssTypographyHeadline } from '../stylex';
import { useCssScope } from '../config-provider/hooks';
import InfoFilled from '@universe-design/icons-react/InfoFilled';
import SucceedFilled from '@universe-design/icons-react/SucceedFilled';
import WarningFilled from '@universe-design/icons-react/WarningFilled';
import ErrorFilled from '@universe-design/icons-react/ErrorFilled';
import type { LiteralUnion } from '../util';
import type { BaseTooltipProps, TooltipProps } from '../tooltip/Tooltip';
import { Tooltip } from '../tooltip/Tooltip';
import type { RenderFunction } from '../util/getRenderPropValue';
import { getRenderPropValue } from '../util/getRenderPropValue';
import { FirstLineAlign } from '../shared/components/FirstLineAlign';
import type { PopperRef } from '../shared/components/popper';

import './style';

// 正常、宽松、无任何内容（并且无填充）
export type PopoverMode = 'normal' | 'loose' | 'pure';
export type PopoverIconType = 'success' | 'warning' | 'info' | 'error';

/**
 * @public
 * @docgen
 */
export interface PopoverProps extends BaseTooltipProps, Pick<TooltipProps, 'offset'> {
  /**
   * 卡片标题图标
   */
  icon?: LiteralUnion<PopoverIconType, ReactNode>;
  /**
   * 卡片标题
   */
  title?: React.ReactNode | RenderFunction;
  /**
   * 卡片内容
   */
  content?: React.ReactNode | RenderFunction;
  /**
   * 卡片模式，`normal` 为正常边距，`loose` 为小边距，`pure` 为无边距
   * @defaultValue `normal`
   */
  mode?: PopoverMode;
}

const iconMap = {
  success: SucceedFilled,
  warning: WarningFilled,
  info: InfoFilled,
  error: ErrorFilled,
};

export const Popover = forwardRef<PopperRef, PopoverProps>(function Popover(props, ref) {
  const {
    mode = 'normal',
    prefixCls,
    title,
    content,
    overlayClassName,
    icon,
    mouseLeaveDelay = 100,
    ...restProps
  } = props;
  const cssScope = useCssScope('popover', prefixCls);
  const contentCls = cssScope('__overlay-content');

  const titleBeforeProps = useMemo<React.ComponentPropsWithoutRef<'div'> | undefined>(() => {
    if (!icon) {
      return;
    }
    const cssIconWrapper = stylex.create({
      paddingRight: 8,
    });
    const cssIcon = stylex.create({
      fontSize: 16,
      lineHeight: 0,
      alignSelf: 'center',
    });
    if (typeof icon === 'string' && (icon === 'success' || icon === 'warning' || icon === 'info' || icon === 'error')) {
      const Icon = iconMap[icon as PopoverIconType];
      return {
        className: `${cssScope()}-icon-${icon as PopoverIconType}`,
        sx: [cssIconWrapper],
        children: <Icon className={stylex(cssIcon)} />,
      };
    }
    return {
      className: `${cssScope()}-icon-customer`,
      sx: [cssIconWrapper],
      children: <span className={clsx(`${cssScope()}-icon-customer-wrapper`, stylex(cssIcon))}>{icon}</span>,
    };
  }, [cssScope, icon]);

  const popupNode = useMemo(() => {
    if (!title && !content) {
      return null;
    }

    // 纯净模式只渲染 content，并且移除了 padding 相关的 style
    if (mode === 'pure') {
      return content && <div className={contentCls}>{getRenderPropValue(content)}</div>;
    }

    return (
      <>
        {title && (
          <FirstLineAlign
            className={cssScope('__overlay-title')}
            sx={[
              cssFlexBase,
              cssTypographyHeadline,
              stylex.create({
                alignItems: 'baseline',
              }),
            ]}
            beforeProps={titleBeforeProps}
          >
            {getRenderPropValue(title)}
          </FirstLineAlign>
        )}
        {content && <div className={contentCls}>{getRenderPropValue(content)}</div>}
      </>
    );
  }, [content, contentCls, cssScope, mode, title, titleBeforeProps]);

  return (
    <Tooltip
      ref={ref}
      prefixCls={cssScope()}
      overlay={popupNode}
      overlayClassName={clsx(overlayClassName, cssScope(`--mode-${mode}`))}
      mouseLeaveDelay={mouseLeaveDelay}
      arrowHasStroke
      {...restProps}
    />
  );
});

if (__DEV__) {
  Popover.displayName = 'Popover';
}
