// @universe-design/react/es/date-picker/RangePicker/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React from 'react';
import { RangePickerTrigger } from './RangePickerTrigger';
import { RangePickerPanel } from './RangePickerPanel';
import { RangePickerContextProvider } from './Context';
import { useRangePicker } from './useRangePicker';
import { RangePickerFooter } from './RangePickerFooter';
import '../style';
export const RangePickerProvider = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    children,
    ...props
  } = _ref;
  const rangePickerCtx = useRangePicker(props);
  React.useImperativeHandle(ref, () => rangePickerCtx);
  return /*#__PURE__*/React.createElement(RangePickerContextProvider, {
    value: rangePickerCtx
  }, children);
});
export const RangePicker = props => {
  const {
    dateRender,
    startOfWeek,
    hideWeekNumbers,
    janX,
    ranges,
    renderExtraFooter,
    ...restProps
  } = props;
  return /*#__PURE__*/React.createElement(RangePickerProvider, restProps, /*#__PURE__*/React.createElement(RangePickerTrigger, _extends({}, props, {
    panel: /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(RangePickerPanel, {
      dateRender: dateRender,
      startOfWeek: startOfWeek,
      hideWeekNumbers: hideWeekNumbers,
      janX: janX
    }), /*#__PURE__*/React.createElement(RangePickerFooter, {
      ranges: ranges,
      renderExtraFooter: renderExtraFooter
    }))
  })));
};
