// @universe-design/react/es/date-picker/RangePicker/RangePickerFooter.js
import clsx from 'clsx';
import React from 'react';
import { Button } from '../../button';
import { useCssScope } from '../../config-provider/hooks';
import { Space } from '../../space';
import { isFunction } from '../../util';
import { useRangePickerContext } from '../RangePicker/Context';
import '../style';
export const RangePickerFooter = props => {
  const {
    renderExtraFooter,
    ranges
  } = props;
  const pickerFooterCssScope = useCssScope('picker__footer');
  const rangePickerCtx = useRangePickerContext();
  const {
    localMode,
    value,
    setValue,
    setOpened
  } = rangePickerCtx;
  const footerContent = [];
  const onQuickSelect = selectedValue => {
    setValue(selectedValue);
    setOpened(false);
  };
  if (renderExtraFooter) {
    footerContent.push(/*#__PURE__*/React.createElement("div", {
      key: "extra",
      className: pickerFooterCssScope('__extra')
    }, renderExtraFooter(rangePickerCtx)));
  } else if (localMode === 'date' && ranges) {
    footerContent.push(/*#__PURE__*/React.createElement("div", {
      key: "default",
      className: pickerFooterCssScope('__range')
    }, /*#__PURE__*/React.createElement(Space, {
      size: "small",
      wrap: true
    }, Object.entries(ranges).map(_ref => {
      var _value$, _value$2;
      let [btnText, rangeValueOrFunc] = _ref;
      let rangeValue;
      if (isFunction(rangeValueOrFunc)) {
        rangeValue = rangeValueOrFunc();
      } else {
        rangeValue = rangeValueOrFunc;
      }
      const isSelected = (value === null || value === void 0 ? void 0 : (_value$ = value[0]) === null || _value$ === void 0 ? void 0 : _value$.toString()) === rangeValue[0].toString() && (value === null || value === void 0 ? void 0 : (_value$2 = value[1]) === null || _value$2 === void 0 ? void 0 : _value$2.toString()) === rangeValue[1].toString();
      return /*#__PURE__*/React.createElement(Button, {
        className: clsx(pickerFooterCssScope('__range__preset-btn')),
        key: btnText,
        type: "text",
        selected: isSelected,
        color: "primary",
        onClick: () => onQuickSelect(rangeValue)
      }, btnText);
    })), /*#__PURE__*/React.createElement("div", {
      style: {
        flex: 1
      }
    })));
  }
  if (footerContent.length) {
    return /*#__PURE__*/React.createElement("div", {
      className: pickerFooterCssScope()
    }, footerContent);
  }
  return null;
};
