// @universe-design/react/es/date-picker/RangePicker/RangePickerTrigger.js
import * as React from 'react';
import CalendarOutlined from '@universe-design/icons-react/CalendarOutlined';
import clsx from 'clsx';
import noop from 'lodash/noop';
import { useComponentSize, useInputGroupSize, useDirection } from '../../config-provider/hooks';
import { isFunction, getDefaultSizeAbbr, KeyCode } from '../../util';
import { KeyboardKey } from '@universe-design/react-utilities';
import { useRangePickerContext } from './Context';
import composeRefs from '../../util/composeRefs';
import { Phase, TriggerType } from './types';
import NativeInput from '../../shared/components/Input';
import { ListBox } from '../../shared/components/list/List';
import { isInRangeView } from '../utils/ui';
import { useComplementEvent } from '../../hooks/useComplementEvent';
import { contains } from '../../util/dom';
import { usePopperSidecar } from '../../config-provider/hooks/usePopperSidecar';
import { SlideUpDownTransition } from '../../shared/components/transition/SlideUpDownTransition';
import InputWrapper from '../../shared/components/picker/Comp/input-wrapper';
import '../style';
import { usePopupPortalContainer } from '@universe-design/react-provider';
var InputType = /*#__PURE__*/function (InputType) {
  InputType[InputType["Start"] = 0] = "Start";
  InputType[InputType["End"] = 1] = "End";
  return InputType;
}(InputType || {});
export function useRangePickerTrigger(props) {
  const {
    onMouseUp,
    onKeyDown,
    inputReadOnly,
    placeholder: placeholderProp,
    autoFocus
  } = props;
  const {
    setOpened,
    startHoverText,
    endHoverText,
    viewValueRef,
    locale,
    textToDate,
    setViewValue,
    prefixCls,
    commitChange,
    disabled,
    picker,
    startText,
    setStartText,
    endText,
    setEndText,
    mode,
    defaultPickerValue,
    startPickerValue,
    endPickerValue,
    setStartPickerValue,
    setEndPickerValue,
    phaseRef,
    setPhase,
    setTriggerType
  } = useRangePickerContext();

  /**
   * Placeholder 初始值
   */
  const placeholder = (() => {
    if (placeholderProp) {
      return placeholderProp;
    }
    if (picker === 'date') {
      return locale.rangePlaceholder;
    }
    if (picker === 'month') {
      return locale.rangeMonthPlaceholder;
    }
    if (picker === 'year') {
      return locale.rangeYearPlaceholder;
    }
    return [];
  })();

  /**
   * Input Event
   */
  const startInputRef = React.useRef(null);
  const endInputRef = React.useRef(null);

  // 当前 focus 在哪个 input
  const focusedInputRef = React.useRef();
  const focusInput = React.useCallback(() => {
    if (phaseRef.current === Phase.EndDate) {
      var _endInputRef$current;
      (_endInputRef$current = endInputRef.current) === null || _endInputRef$current === void 0 ? void 0 : _endInputRef$current.focus();
    } else {
      var _startInputRef$curren2;
      (_startInputRef$curren2 = startInputRef.current) === null || _startInputRef$curren2 === void 0 ? void 0 : _startInputRef$curren2.focus();
    }
  }, [endInputRef, phaseRef, startInputRef]);
  return {
    handleInputWrapperMouseUp: function (e) {
      if (e.isDefaultPrevented()) {
        return;
      }
      isFunction(onMouseUp) && onMouseUp(e);
      if (e.isDefaultPrevented()) {
        return;
      }
      if (!open) {
        var _startInputRef$curren;
        setOpened(true);
        (_startInputRef$curren = startInputRef.current) === null || _startInputRef$curren === void 0 ? void 0 : _startInputRef$curren.focus();
      }
    },
    getInputProps: triggerType => {
      const isStart = triggerType === TriggerType.StartDate;
      const inputRef = isStart ? startInputRef : endInputRef;
      const triggerName = isStart ? 'start' : 'end';
      const resetPickerValue = () => {
        const index = isStart ? 0 : 1;
        const target = viewValueRef.current[index] || defaultPickerValue[index];
        if (!target) {
          return;
        }

        // 如果值在当前面板中，不跳转
        if (isInRangeView(target, [startPickerValue, endPickerValue], mode)) {
          return;
        }

        // 如果当前值不在面板中，跳转到该值对应面板上
        if (isStart) {
          setStartPickerValue(target);
        } else {
          setEndPickerValue(target);
        }
      };
      return {
        size: 12,
        value: isStart ? startHoverText || startText : endHoverText || endText,
        onChange: e => {
          const date = textToDate(e.target.value);

          // 不管怎么样都先更新 input 里面的 value
          if (triggerType === TriggerType.StartDate) {
            setStartText(e.target.value);
          }
          if (triggerType === TriggerType.EndDate) {
            setEndText(e.target.value);
          }
          if (date && date.isValid()) {
            const [start, end] = viewValueRef.current;
            if (triggerType === TriggerType.StartDate) {
              // input 输入的时候出现了有效值但是并不会立即提交值
              setViewValue([date, end]);
            } else {
              setViewValue([start, date]);
            }
            resetPickerValue();
          }
        },
        onFocus: () => {
          setTriggerType(triggerType);
          // 保证 focus 和 phase 同步
          const targetPhase = isStart ? Phase.StartDate : Phase.EndDate;
          setPhase(targetPhase);
          const currentFocused = isStart ? InputType.Start : InputType.End;
          if (focusedInputRef.current !== currentFocused) {
            /**
             * 对于 从panel 重新 focus 回 input 的场景，不 resetPickerValue
             * 因为可能在面板中修改了年份或月份
             */
            resetPickerValue();
            focusedInputRef.current = currentFocused;
          }
        },
        onClick: ev => {
          var _ev$currentTarget;
          // 为了点击的时候选中 input 里的内容
          (_ev$currentTarget = ev.currentTarget) === null || _ev$currentTarget === void 0 ? void 0 : _ev$currentTarget.select();
        },
        onMouseDown: () => {
          setOpened(true);
        },
        className: clsx({
          [`${prefixCls}-input`]: true,
          [`${prefixCls}-input-${triggerName}`]: true,
          [`${prefixCls}-input-placeholder`]: Boolean(isStart ? startHoverText : endHoverText)
        }),
        disabled: isStart ? disabled[0] : disabled[1],
        autoFocus: isStart ? autoFocus : undefined,
        ref: inputRef,
        onKeyDown: e => {
          isFunction(onKeyDown) && onKeyDown(e);
          if (e.key !== KeyboardKey.Enter) {
            setOpened(true);
            return;
          }
          if (e.keyCode === KeyCode.ENTER) {
            commitChange();
            setOpened(false);
          }
        },
        // TODO 之前是这样的 inputReadOnly || (isFunction(format) && !isFunction(parse))，不知道后面那坨是做什么的先暂时去掉了
        readOnly: inputReadOnly,
        placeholder: isStart ? placeholder[0] : placeholder[1]
      };
    },
    startText,
    endText,
    focusInput
  };
}
export function RangePickerTrigger(props) {
  const defaultSize = useComponentSize();
  const inputGroupSize = useInputGroupSize();
  const {
    size,
    error,
    allowClear,
    className,
    style,
    suffixIcon = /*#__PURE__*/React.createElement(CalendarOutlined, null),
    onMouseDown = noop,
    onClick = noop,
    dropdownClassName,
    getPopupContainer,
    triggerElement,
    popupStyle,
    panel
  } = props;
  const direction = useDirection();
  const Popper = usePopperSidecar();
  const cmpSize = getDefaultSizeAbbr(size || inputGroupSize || defaultSize);
  const panelDivRef = React.useRef(null);
  const inputDivRef = React.useRef(null);
  const triggerElementRef = React.useRef(null);
  const rangePickerCtx = useRangePickerContext();
  const {
    triggerType,
    open,
    setOpened,
    startHoverText,
    endHoverText,
    handleClear,
    value,
    handlePopupVisibleChange,
    startViewDate,
    endViewDate,
    triggerElementEnabled,
    cssScope,
    prefixCls,
    disabled,
    commitChange
  } = rangePickerCtx;

  /**
   * 控制 popup 在点击非内部元素时自动关闭并提交数据
   */
  useComplementEvent('mousedown', {
    listener: React.useCallback(() => {
      commitChange();
      setOpened(false);
    }, [commitChange, setOpened]),
    elements: React.useMemo(() => [panelDivRef, inputDivRef, triggerElementRef], [])
  });
  const {
    getInputProps,
    startText,
    endText,
    handleInputWrapperMouseUp,
    focusInput
  } = useRangePickerTrigger(props);
  const rootCls = clsx(cssScope(), {
    [cssScope(`--size-${cmpSize}`)]: cmpSize,
    [cssScope('--error')]: error,
    [cssScope('--focus')]: triggerType !== TriggerType.None
  });
  const ctxGetPopupContainer = usePopupPortalContainer();
  let triggerElementNode;
  if (isFunction(triggerElement)) {
    triggerElementNode = triggerElement({
      startText,
      endText,
      startHoverText,
      endHoverText,
      value,
      open
    });
  } else {
    triggerElementNode = triggerElement;
  }

  /**
   * 用于保证点击面板之后， input 被 focus 回去
   * 同时也需要保证 面板里的 input 能被 focus
   */
  const onPanelClick = React.useCallback(e => {
    const {
      target
    } = e;
    const popupEle = panelDivRef.current;
    // 如果当前 focus 的不是 popup 内部的元素，就聚焦 input
    window.setTimeout(() => {
      // 如果焦点在面板上, 不 focus 到 input 上
      if (document.activeElement && (contains(popupEle, document.activeElement) || contains(target, document.activeElement))) {
        return;
      }
      // 其他情况 focus input
      focusInput();
    });
  }, [focusInput]);
  return /*#__PURE__*/React.createElement(Popper, {
    destroyOnHide: true,
    visible: open,
    action: undefined,
    showAction: triggerElementEnabled ? ['click'] : [],
    hideAction: [],
    popupClassName: clsx(dropdownClassName, `${prefixCls}-dropdown`),
    getPopupContainer: triggerNode => {
      const _getPopupContainer = getPopupContainer !== null && getPopupContainer !== void 0 ? getPopupContainer : ctxGetPopupContainer;
      return _getPopupContainer(triggerNode);
    },
    onVisibleChange: handlePopupVisibleChange,
    popupStyle: popupStyle,
    popup: /*#__PURE__*/React.createElement(ListBox, {
      className: `${prefixCls}__list__box`,
      onClick: onPanelClick,
      ref: panelDivRef
    }, panel),
    prefixCls: prefixCls,
    popupMotion: "slide-up",
    popupPlacement: direction === 'rtl' ? 'bottomRight' : 'bottomLeft',
    enableRestoreFocus: false,
    initialFocus: null,
    showArrow: false,
    arrowHasStroke: false,
    Transition: SlideUpDownTransition,
    enableMaxHeight: false,
    defaultMainAxisOffsetWithoutArrow: 4
  }, (() => {
    if (triggerElementNode) {
      return /*#__PURE__*/React.cloneElement(triggerElementNode, {
        ref: composeRefs(triggerElementRef, triggerElementNode.ref)
      });
    }
    return /*#__PURE__*/React.createElement("div", {
      style: style,
      ref: inputDivRef,
      className: rootCls
    }, /*#__PURE__*/React.createElement(InputWrapper, {
      className: clsx(className, `${prefixCls}-rangeInput`),
      onClear: handleClear,
      prefixCls: prefixCls,
      onMouseUp: handleInputWrapperMouseUp,
      onMouseDown: onMouseDown,
      onClick: onClick,
      focused: triggerType !== TriggerType.None,
      suffixIcon: suffixIcon,
      disabled: disabled[0] && disabled[1],
      showClear: Boolean(startViewDate) && Boolean(endViewDate),
      allowClear: allowClear
    }, /*#__PURE__*/React.createElement(NativeInput, getInputProps(TriggerType.StartDate)), /*#__PURE__*/React.createElement("div", {
      className: `${prefixCls}-rangeInput-seperator`
    }, /*#__PURE__*/React.createElement("div", {
      className: `${prefixCls}-seperator`
    })), /*#__PURE__*/React.createElement(NativeInput, getInputProps(TriggerType.EndDate))));
  })());
}
