// @universe-design/react/es/date-picker/RangePicker/range-picker-strategies.js
import { Phase, TriggerType } from './types';
const rangePickerSelectionStrategyStartDate = {
  name: 'StartDate',
  defaultPhase: Phase.StartDate,
  useHooks: _ref => {
    let {
      phaseRef,
      viewValueRef,
      mode,
      setPhase,
      setViewValue,
      setOpened
    } = _ref;
    return {
      onDateSelect: date => {
        const currentPhase = phaseRef.current;
        const [start, end] = viewValueRef.current;
        // 选择开始时间
        if (currentPhase === Phase.StartDate) {
          // 需要重新选取结束时间的情况
          // 1. 如果选择时间在结束时间之后，说明结束时间失效了
          // 2. 如果没有选取过结束时间
          if (date.isAfter(end, mode) || end === undefined) {
            setPhase(Phase.EndDate);
            setViewValue([date, undefined]);
            return;
          }
          // 对于合法的情况，直接填入选择值，并更新给用户。
          setViewValue([date, end], {
            autoCommit: true
          });
          setOpened(false);
          return;
        }
        // 选择结束时间
        if (currentPhase === Phase.EndDate) {
          // 如果选中的结束时间比开始时间早就认为是重置开始时间
          if (date.isBefore(start, mode)) {
            setPhase(Phase.EndDate);
            setViewValue([date, undefined]);
            return Phase.StartDate;
          }
          // 选择完毕以后直接提交退出
          setViewValue([start, date], {
            autoCommit: true
          });
          setOpened(false);
        }
      },
      onPopupVisibleChange: visible => {
        if (visible) {
          setPhase(Phase.StartDate);
          setOpened(true, TriggerType.StartDate);
        } else {
          setOpened(false);
        }
      }
    };
  }
};
const rangePickerSelectionStrategyEndDate = {
  name: 'EndDate',
  defaultPhase: Phase.EndDate,
  useHooks: _ref2 => {
    let {
      setPhase,
      phaseRef,
      viewValueRef,
      setViewValue,
      mode,
      setOpened
    } = _ref2;
    return {
      onDateSelect: date => {
        const currentPhase = phaseRef.current;
        const [start, end] = viewValueRef.current;
        if (currentPhase === Phase.EndDate) {
          if (date.isBefore(start, mode) || start === undefined) {
            setPhase(Phase.StartDate);
            setViewValue([undefined, date]);
            return;
          }
          setViewValue([start, date], {
            autoCommit: true
          });
          setOpened(false);
          return;
        }
        if (currentPhase === Phase.StartDate) {
          if (date.isAfter(end, mode)) {
            setPhase(Phase.StartDate);
            setViewValue([undefined, date]);
            return Phase.EndDate;
          }
          setViewValue([date, end], {
            autoCommit: true
          });
          setOpened(false);
        }
      },
      onPopupVisibleChange: visible => {
        if (visible) {
          setPhase(Phase.EndDate);
          setOpened(true, TriggerType.EndDate);
        } else {
          setOpened(false);
        }
      }
    };
  }
};
const rangePickerSelectionStrategyStartDateOverall = {
  name: 'StartDateOverall',
  defaultPhase: Phase.StartDate,
  useHooks: _ref3 => {
    let {
      setPhase,
      phaseRef,
      viewValueRef,
      setViewValue,
      mode,
      setOpened
    } = _ref3;
    return {
      onDateSelect: date => {
        /**
         * 规则
         * 1. 只要是合法的值就 commit
         * 2. 每次都完整进行 start -> end 的操作
         */
        const currentPhase = phaseRef.current;
        const [start, end] = viewValueRef.current;
        if (currentPhase === Phase.StartDate) {
          if (date.isAfter(end, mode) || end === undefined) {
            // 不合法直接取消之前的结束时间
            setPhase(Phase.EndDate);
            setViewValue([date, undefined]);
            return;
          }
          setPhase(Phase.EndDate);
          setViewValue([date, end], {
            autoCommit: true
          });
          return;
        }
        if (currentPhase === Phase.EndDate) {
          if (!start) {
            setPhase(Phase.StartDate);
            setViewValue([undefined, date]);
            return;
          }
          if (date.isBefore(start, mode)) {
            setPhase(Phase.EndDate);
            setViewValue([date, undefined]);
            return Phase.StartDate;
          }
          // 提交给用户
          setPhase(Phase.StartDate);
          setViewValue([start, date], {
            autoCommit: true
          });
          setOpened(false);
        }
      },
      onPopupVisibleChange: visible => {
        if (visible) {
          setPhase(Phase.StartDate);
          setOpened(true);
        } else {
          setOpened(false);
        }
      }
    };
  }
};
export { rangePickerSelectionStrategyStartDate, rangePickerSelectionStrategyEndDate, rangePickerSelectionStrategyStartDateOverall };
