// @universe-design/react/es/date-picker/RangePicker/RangePickerPanel.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.u18og1rn{margin-top:15px;}
import uuf62k4 from "./../../.stylex/uuf62k4";
//.u1g777gl{margin-right:15px;}
import ux5slyd from "./../../.stylex/ux5slyd";
//.u6jp0wq{margin-bottom:15px;}
import u1ntxetr from "./../../.stylex/u1ntxetr";
//.u1abw64p{margin-left:15px;}
import u1jmgjb9 from "./../../.stylex/u1jmgjb9";
import React, { useState } from 'react';
import clsx from 'clsx';
import { Space } from '../../space';
import { getHigherLevelMode, getNext, getPrev } from '../panels/base/utils';
import { useRangePickerContext } from './Context';
import RangePickerHeader from './RangePickerHeader';
import { NextMonthSwitcher, NextYearSwitcher, Next20YearsSwitcher, PreMonthSwitcher, PreYearSwitcher, Pre20YearsSwitcher } from '../panels/PanelHeaderSwitchers';
import { MonthYearSelector } from '../panels/selector/MonthYearSelector';
import stylex from '@universe-design/stylex';
import { sxPickerHeaderSwitchWrapper } from '../sx';
import { DateCalendarUI } from '../panels/DateCalendar/DateCalendarUI';
import '../style';

/**
 * 多层选择器解决方案:
 * 当前mode下面，选择日期，才能生效
 * 其他mode上面选择日期，都只是一个中间态，可以在panel组件内部解决
 */

function getChanged(date, mode) {
  const seq = ['date', 'month', 'year'];
  const index = seq.indexOf(mode);
  if (index < 0) {
    return {};
  }
  return seq.slice(index).reduce((pre, mod) => {
    pre[mod] = date.get(mod);
    return pre;
  }, {});
}
function RangePickerPanel(props) {
  // Props
  const {
    className,
    dateRender,
    startOfWeek,
    hideWeekNumbers,
    janX
  } = props;

  // Context
  const {
    prefixCls,
    viewValue,
    mode,
    handleDateSelect,
    handleDateCellMouseEnter,
    handleDateCellMouseLeave,
    isDisabled,
    isPreviewDate,
    localMode,
    setLocalMode,
    startPickerValue,
    endPickerValue,
    setStartPickerValue,
    setEndPickerValue,
    cssScope,
    panelIntervalLimits,
    maxDate,
    minDate
  } = useRangePickerContext();
  const panelLimited = panelIntervalLimits === 'adjacent';
  const [focusedIndex, setFocusedIndex] = useState(0);
  const handleFocusedPanelSelect = date => {
    const newViewDate = date.clone();
    newViewDate.set(getChanged(date, localMode));
    if (focusedIndex === 1) {
      setEndPickerValue(newViewDate);
    } else {
      setStartPickerValue(newViewDate);
    }
    if (mode !== localMode) {
      setLocalMode(mode);
    }
  };
  const handleCollapsed = index => {
    setFocusedIndex(index);
    const nextMode = getHigherLevelMode(mode);
    if (nextMode) {
      setLocalMode(nextMode);
    }
  };
  const handleMonthClick = index => {
    setFocusedIndex(index);
    setLocalMode('month');
  };
  const handleYearClick = index => {
    setFocusedIndex(index);
    setLocalMode('year');
  };
  if (localMode !== mode) {
    return /*#__PURE__*/React.createElement(DateCalendarUI, {
      sx: [{
        "props": {
          "B6of3ja": uuf62k4,
          "t21cq0": ux5slyd,
          "jrapky": u1ntxetr,
          "Frg6f3": u1jmgjb9
        }
      }],
      key: focusedIndex,
      mode: localMode,
      disabledDate: isDisabled,
      defaultPickerValue: focusedIndex === 0 ? startPickerValue : endPickerValue,
      onSelect: handleFocusedPanelSelect,
      startOfWeek: startOfWeek,
      minDate: minDate,
      maxDate: maxDate,
      hideWeekNumbers: hideWeekNumbers,
      janX: janX
    });
  }
  const getPickerBodyProps = index => {
    const dates = viewValue.filter(x => Boolean(x));
    return {
      disabledDate: isDisabled,
      minDate,
      maxDate,
      previewDate: isPreviewDate,
      pickerValue: index === 0 ? startPickerValue : endPickerValue,
      mode,
      onSelect: handleDateSelect,
      onDateMouseEnter: handleDateCellMouseEnter,
      onContentMouseLeave: handleDateCellMouseLeave,
      startOfWeek,
      hideWeekNumbers,
      janX,
      ...(dateRender ? {
        cellRenderer: (date, _ref) => {
          let {
            viewMode,
            defaultNode
          } = _ref;
          return dateRender(date, viewMode, defaultNode);
        }
      } : {}),
      ...(viewValue[0] && viewValue[1] ? {
        selectedRange: viewValue
      } : {
        selectedDates: dates
      })
    };
  };
  const getMonthYearSelectorProps = index => {
    const MonthYearSelectorCls = cssScope('-panel-header-content__center');
    const resValue = index === 0 ? startPickerValue : endPickerValue;
    return {
      mode,
      pickerValue: resValue,
      className: MonthYearSelectorCls,
      hiddenCollapse: mode === 'year',
      collapsed: localMode === mode,
      onMonthClick: () => handleMonthClick(index),
      onYearClick: () => handleYearClick(index),
      onCollapsed: () => handleCollapsed(index)
    };
  };
  const isDate = mode === 'date';
  const isYear = mode === 'year';

  // TODO handleFns
  const pickerPanelRenderer = index => {
    const isLeft = index === 0;
    const isRight = index === 1;
    const pickerValue = isLeft ? startPickerValue.clone() : endPickerValue.clone();
    const switchPicker = value => {
      if (isLeft) {
        setStartPickerValue(value.clone());
      } else {
        setEndPickerValue(value.clone());
      }
    };

    // handleTimeFns
    const goPrevMonth = () => switchPicker(getPrev(pickerValue, 'date'));
    const goNextMonth = () => switchPicker(getNext(pickerValue, 'date'));
    const goPrevYear = () => switchPicker(getPrev(pickerValue, 'month'));
    const goNextYear = () => switchPicker(getNext(pickerValue, 'month'));
    const goPre20Years = () => switchPicker(getPrev(pickerValue, 'year'));
    const goNext20Years = () => switchPicker(getNext(pickerValue, 'year'));
    return /*#__PURE__*/React.createElement(DateCalendarUI, _extends({
      headerRenderer: () => /*#__PURE__*/React.createElement(RangePickerHeader, null, /*#__PURE__*/React.createElement(Space, {
        size: 4,
        className: stylex(sxPickerHeaderSwitchWrapper)
      }, isYear && /*#__PURE__*/React.createElement(Pre20YearsSwitcher, {
        asPlaceholder: isRight && panelLimited,
        onClick: goPre20Years
      }), !isYear && /*#__PURE__*/React.createElement(PreYearSwitcher, {
        asPlaceholder: isRight && panelLimited,
        onClick: goPrevYear
      }), isDate && /*#__PURE__*/React.createElement(PreMonthSwitcher, {
        asPlaceholder: isRight && panelLimited,
        onClick: goPrevMonth
      })), /*#__PURE__*/React.createElement(MonthYearSelector, getMonthYearSelectorProps(index)), /*#__PURE__*/React.createElement(Space, {
        size: 4,
        className: stylex(sxPickerHeaderSwitchWrapper)
      }, isDate && /*#__PURE__*/React.createElement(NextMonthSwitcher, {
        asPlaceholder: isLeft && panelLimited,
        onClick: goNextMonth
      }), !isYear && /*#__PURE__*/React.createElement(NextYearSwitcher, {
        asPlaceholder: isLeft && panelLimited,
        onClick: goNextYear
      }), isYear && /*#__PURE__*/React.createElement(Next20YearsSwitcher, {
        asPlaceholder: isLeft && panelLimited,
        onClick: goNext20Years
      })))
    }, getPickerBodyProps(index)));
  };
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("div", {
    className: clsx(className, `${prefixCls}-panels`)
  }, pickerPanelRenderer(0), pickerPanelRenderer(1)));
}
export { RangePickerPanel };
