// @universe-design/react/es/date-picker/RangePicker/useRangePicker.js
import * as React from 'react';
import { useControlledState, useLatestState } from '@universe-design/react-utilities';
import { useCssScope, useLocaleReceiver } from '../../config-provider/hooks';
import useSyncState from '../../hooks/useSyncState';
import { isFunction } from '../../util';
import useFormatAndParser from '../hooks/useFormatAndParser';
import { isSame, isValidRangeValue } from '../utils/date';
import { isInHoverRange, isInRangeView } from '../utils/ui';
import { rangePickerSelectionStrategyEndDate, rangePickerSelectionStrategyStartDate, rangePickerSelectionStrategyStartDateOverall } from './range-picker-strategies';
import { Phase, TriggerType } from './types';
import usePickerValue from './usePickerValue';
import { isInputDateDisabled } from '../utils/disabled';
const EMPTY_RANGE_VALUE = [];
const isRangeValueEqual = (a, b) => isSame(a[0], b[0]) && isSame(a[1], b[1]);
function useRangePickerCssScope() {
  const cssScope = useCssScope('picker');
  const prefixCls = cssScope();
  return {
    cssScope,
    prefixCls
  };
}
export function useRangePicker(props) {
  const locale = useLocaleReceiver('datePicker');
  const {
    cssScope,
    prefixCls
  } = useRangePickerCssScope();
  const {
    triggerElement,
    defaultOpen = false,
    onOpenChange,
    open: openProp,
    value: valueProp,
    defaultValue = EMPTY_RANGE_VALUE,
    onChange,
    format,
    picker = 'date',
    mode = picker,
    onSelect,
    onClear,
    disabled = [],
    selectionStrategy,
    // ranges,
    // renderExtraFooter,
    disabledDate,
    minDate,
    maxDate,
    // startOfWeek = 0,
    // dateRender,
    defaultPickerValue: defaultPickerValueProp,
    panelIntervalLimits = 'adjacent',
    parse
  } = props;
  const {
    formatValue,
    parseValue
  } = useFormatAndParser({
    mode,
    format,
    parse,
    // rangePicker 不支持 showTime
    showTime: false
  });
  const dateToText = React.useCallback(date => formatValue(date), [formatValue]);
  const textToDate = React.useCallback(text => parseValue(text), [parseValue]);

  /**
   * range value 的受控值
   */
  const [value, setValue] = useControlledState(defaultValue, valueProp !== null && valueProp !== void 0 ? valueProp : EMPTY_RANGE_VALUE, {
    onChange,
    isEqual: isRangeValueEqual,
    isControlledOverride: 'value' in props
  });

  /**
   * view value 的半受控值，显示的选中范围可能会和真实的范围值不同
   */
  const [viewValue = EMPTY_RANGE_VALUE, _setViewValue, viewValueRef] = useSyncState(value, {
    isEqual: isRangeValueEqual
  });
  const [startViewDate, endViewDate] = viewValue;

  /**
   * Input 内部显示值的状态，由于可能会在输入的时候和 viewValue 不同，所以用 useSyncValue
   */
  const [startText, setStartText] = useSyncState(dateToText(startViewDate));
  const [endText, setEndText] = useSyncState(dateToText(endViewDate));
  /**
   * 打开状态的受控值
   */
  const [open, _setOpened] = useControlledState(defaultOpen, openProp, {
    onChange: onOpenChange
  });

  // 触发打开 popover 的 action 类型
  const [triggerTypeState, _setTriggerType] = useLatestState(
  // 如果默认 open 的话, 我们认为聚焦类型是未知
  () => open ? TriggerType.Unknown : TriggerType.None);
  const setTriggerType = React.useCallback(triggerType => {
    /**
     * 这里尽力而为地判断是否常驻打开
     * 如果常驻打开的话，永远使用 overrall 的 strategy
     */

    _setTriggerType(openProp && !onOpenChange ? TriggerType.Unknown : triggerType);
  }, [_setTriggerType, onOpenChange, openProp]);

  // 这里保证只切换 hooks 而不影响 hooks 数量和顺序
  const rangePickerSelectionStrategy = selectionStrategy !== null && selectionStrategy !== void 0 ? selectionStrategy :
  // eslint-disable-next-line no-nested-ternary
  triggerTypeState === TriggerType.StartDate ? rangePickerSelectionStrategyStartDate : triggerTypeState === TriggerType.EndDate ? rangePickerSelectionStrategyEndDate : rangePickerSelectionStrategyStartDateOverall;
  // 现在的选择阶段，开始时间、结束时间还是未选中
  const [phase, setPhase, phaseRef] = useLatestState(rangePickerSelectionStrategy.defaultPhase);

  /**
   * 把 viewValue 提交给用户
   */
  const commitChange = React.useCallback(() => {
    const [start, end] = viewValueRef.current;
    if (!isValidRangeValue([start, end]) || start && isInputDateDisabled(start, {
      disabledDate,
      minDate,
      maxDate
    }) || end && isInputDateDisabled(end, {
      disabledDate,
      minDate,
      maxDate
    })) {
      // 非法值，使用 value 重置
      _setViewValue(value);
    } else {
      setValue(viewValueRef.current);
      // 此时强制重置 inputvalue, 防止存在中间态
      setStartText(dateToText(start));
      setEndText(dateToText(end));
      setTriggerType(TriggerType.None);
    }
  }, [viewValueRef, disabledDate, minDate, maxDate, _setViewValue, value, setValue, setStartText, dateToText, setEndText, setTriggerType]);
  const setViewValue = React.useCallback(function (_value) {
    let {
      autoCommit = false
    } = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    _setViewValue(_value);
    if (autoCommit) {
      commitChange();
    }
  }, [_setViewValue, commitChange]);
  const triggerElementEnabled = Boolean(isFunction(triggerElement) || triggerElement);
  const {
    setStartPickerValue,
    setEndPickerValue,
    startPickerValue,
    endPickerValue,
    defaultPickerValue,
    setPickerValue,
    pickerValue
  } = usePickerValue({
    defaultPickerValue: defaultPickerValueProp,
    value: viewValue,
    mode,
    panelIntervalLimited: panelIntervalLimits === 'adjacent'
  });
  const setOpened = React.useCallback((visible, triggerType) => {
    _setOpened(visible);
    if (!visible) {
      // 把半受控的 viewValue 重置
      setViewValue(value);
      setHoverDate(undefined);
    }

    /**
     * 仅用于兼容老 API，内部不在此设置 triggerType
     */
    if (triggerType !== undefined) {
      setTriggerType(triggerType);
    }
  }, [_setOpened, setTriggerType, setViewValue, value]);

  /**
   * 悬停日期
   */
  const [hoverDate, setHoverDate] = React.useState();
  const startHoverText = phase === Phase.StartDate ? dateToText(hoverDate) : '';
  const endHoverText = phase === Phase.EndDate ? dateToText(hoverDate) : '';
  const {
    onDateSelect,
    onPopupVisibleChange
  } = rangePickerSelectionStrategy.useHooks({
    setPhase,
    phaseRef,
    setViewValue,
    commitChange,
    setOpened,
    viewValueRef,
    mode
  });
  const handleClear = React.useCallback(() => {
    isFunction(onClear) && onClear();
    setValue(EMPTY_RANGE_VALUE);
  }, [onClear, setValue]);

  /**
   * 当前面板的选择模式
   */
  const [localMode, setLocalMode] = useSyncState(mode);
  const handleDateSelect = React.useCallback(date => {
    const currentPhase = phaseRef.current;
    // 可能在选开始时间的阶段，选了一个很晚的时间，导致 actualPhase 被重置为 endDate
    const actualPhase = onDateSelect(date);
    const _phase = actualPhase !== null && actualPhase !== void 0 ? actualPhase : currentPhase;
    isFunction(onSelect) && onSelect(date, _phase);

    /**
     * 如果点击到了可视区域之外，切换面板
     *
     * 这里无法通过 value 推导出来，因为需要确定点击时间属于哪个 phase
     **/
    if (!isInRangeView(date, [startPickerValue, endPickerValue], mode)) {
      if (_phase === Phase.StartDate) {
        setStartPickerValue(date);
      } else {
        setEndPickerValue(date);
      }
    }
  }, [startPickerValue, mode, onDateSelect, onSelect, phaseRef, endPickerValue, setStartPickerValue, setEndPickerValue]);
  return {
    // States
    startText,
    setStartText,
    endText,
    setEndText,
    open,
    triggerType: triggerTypeState,
    value,
    viewValue,
    triggerElementEnabled,
    hoverDate,
    phase,
    localMode,
    // Refs,
    viewValueRef,
    phaseRef,
    // Mutations
    setOpened,
    setValue,
    setLocalMode,
    // Actions
    handleDateSelect,
    handleDateCellMouseEnter: setHoverDate,
    handleDateCellMouseLeave: () => setHoverDate(undefined),
    handleClear,
    handlePopupVisibleChange: onPopupVisibleChange,
    commitChange,
    setViewValue,
    // Views
    startHoverText,
    endHoverText,
    startViewDate,
    endViewDate,
    // Others
    isDisabled: date => {
      if (isFunction(disabledDate)) {
        return disabledDate(date);
      }
      // TODO 这里的逻辑先注释掉，因为如果有禁用状态的话会导致“当用户想直接选中结束时间之后的时间作为起始时间”时，用户需要选择两次
      // const [start, end] = viewValue;
      // if (phase === Phase.StartDate) {
      //   return end === undefined ? false : date.isAfter(end, mode);
      // }
      // if (phase === Phase.EndDate) {
      //   return start === undefined ? false : date.isBefore(start, mode);
      // }
      return false;
    },
    minDate,
    maxDate,
    isPreviewDate: date => {
      if (phase === Phase.StartDate) {
        return isInHoverRange(date, mode, hoverDate, viewValue, 0);
      } else if (phase === Phase.EndDate) {
        return isInHoverRange(date, mode, hoverDate, viewValue, 1);
      }
      return false;
    },
    locale,
    dateToText,
    textToDate,
    // Props
    // ranges,
    // renderExtraFooter,
    picker,
    mode,
    cssScope,
    prefixCls,
    disabled,
    defaultPickerValue,
    setPhase,
    setTriggerType,
    // startOfWeek,
    // dateRender,
    startPickerValue,
    endPickerValue,
    setStartPickerValue,
    setEndPickerValue,
    panelIntervalLimits,
    setPickerValue,
    pickerValue
  };
}
