// @universe-design/react/src/menu/internal-menu/Menu.tsx
/* eslint-disable react-hooks/exhaustive-deps */
import * as React from 'react';
import { flushSync } from 'react-dom';
import type { CSSMotionProps } from '@universe-design/rc-motion';
import clsx from 'clsx';
import { shallowEqualObjects as shallowEqual } from 'shallow-equal';
import useMergedState from '@universe-design/rc-util/lib/hooks/useMergedState';
import warning from '@universe-design/rc-util/lib/warning';
import Overflow from 'rc-overflow';
import type {
  BuiltinPlacements,
  MenuClickEventHandler,
  RcMenuInfo,
  RcMenuMode,
  MenuSelectEventHandler,
  TriggerSubMenuAction,
  MenuSelectInfo,
  RenderIconType,
} from './interface';
import RcMenuItem from './MenuItem';
import { parseChildren } from './utils/nodeUtil';
import MenuContextProvider from './context/MenuContext';
import useMemoCallback from './hooks/useMemoCallback';
import { warnItemProp } from './utils/warnUtil';
import RcSubMenu from './SubMenu';
import useAccessibility from './hooks/useAccessibility';
import useUUID from './hooks/useUUID';
import { PathRegisterContext, PathUserContext } from './context/PathContext';
import useKeyRecords, { OVERFLOW_KEY } from './hooks/useKeyRecords';
import { IdContext } from './context/IdContext';
import { useCssScope } from '../../config-provider';
import { navMenuDefaults } from '../constants';

/**
 * Menu modify after refactor:
 * ## Add
 * - disabled
 *
 * ## Remove
 * - openTransitionName
 * - openAnimation
 * - onDestroy
 * - siderCollapsed: Seems antd do not use this prop (Need test in antd)
 * - collapsedWidth: Seems this logic should be handle by antd Layout.Sider
 */

// optimize for render
const EMPTY_LIST: string[] = [];

export interface RcMenuProps extends Omit<React.HTMLAttributes<HTMLUListElement>, 'onClick' | 'onSelect' | 'dir'> {
  prefixCls?: string;

  children?: React.ReactNode;
  /**
   * 是否禁用
   */
  disabled?: boolean;
  /** @private Disable auto overflow. Pls note the prop name may refactor since we do not final decided. */
  disabledOverflow?: boolean;

  /**
   *  direction of menu
   *  @internal
   **/
  direction?: 'ltr' | 'rtl';

  // Mode
  mode?: RcMenuMode;
  /**
   * `inline`时菜单是否收起状态
   */
  inlineCollapsed?: boolean;
  /**
   * 初始展开的 SubMenu 菜单项`key`数组
   */
  defaultOpenKeys?: string[];
  /**
   * 当前展开的 SubMenu 菜单项`key`数组
   */
  openKeys?: string[];
  /**
   * @private
   **/
  activeKey?: string;
  /**
   * @private
   **/
  defaultActiveFirst?: boolean;
  /**
   * 是否允许选中
   * @defaultValue true
   */
  selectable?: boolean;
  /**
   * 是否允许多选
   * @defaultValue false
   */
  multiple?: boolean;
  /**
   * 初始选中的菜单项`key`数组
   */
  defaultSelectedKeys?: string[];
  /**
   * 当前选中的菜单项`key`数组
   */
  selectedKeys?: string[];
  /**
   * 被选中时调用
   */
  onSelect?: MenuSelectEventHandler;
  /**
   * 取消选中时调用
   */
  onDeselect?: MenuSelectEventHandler;
  /**
   * `inline`时菜单的缩进宽度
   */
  inlineIndent?: number;

  // Motion
  /**
   * Menu motion define. Use `defaultMotions` if you need config motion of each mode
   * @private
   * @internal
   **/
  motion?: CSSMotionProps;
  /**
   * @private
   * @internal
   * Default menu motion of each mode
   **/
  defaultMotions?: Partial<{ [key in RcMenuMode | 'other']: CSSMotionProps }>;
  /**
   * 用户鼠标进入子菜单后开启延时，单位为毫秒
   * @defaultValue 0
   */
  subMenuOpenDelay?: number;
  /**
   * 用户鼠标离开子菜单后关闭延时，单位为毫秒
   * @defaultValue 100
   */
  subMenuCloseDelay?: number;
  /**
   * 在子菜单展示之前就渲染进 DOM
   * @defaultValue false
   */
  forceSubMenuRender?: boolean;
  /**
   * SubMenu 展开/关闭的触发行为 ( mode = "inline" 模式下仅支持 click )
   * @defaultValue hover
   */
  triggerSubMenuAction?: TriggerSubMenuAction;
  builtinPlacements?: BuiltinPlacements;

  // Icon
  itemIcon?: RenderIconType;
  /**
   * 自定义次级菜单展开图标
   */
  expandIcon?: RenderIconType;
  /**
   * 自定义 Menu 折叠时的图标
   */
  overflowedIndicator?: React.ReactNode;

  /**
   * 自定义弹出容器
   */
  getPopupContainer?: (node: HTMLElement) => HTMLElement;

  /**
   * @private
   */
  onClick?: MenuClickEventHandler;
  /**
   * SubMenu 展开/关闭的回调
   */
  onOpenChange?: (openKeys: React.Key[]) => void;
}

interface LegacyMenuProps extends RcMenuProps {
  openTransitionName?: string;
  openAnimation?: string;
}

const RcMenu: React.FC<RcMenuProps> = props => {
  const cssScope = useCssScope('internal-menu');
  const {
    prefixCls = cssScope(),
    style,
    className,
    tabIndex = 0,
    children,
    direction,

    id,

    // Mode
    mode = 'vertical',
    inlineCollapsed,

    // Disabled
    disabled,
    disabledOverflow,

    // Open
    subMenuOpenDelay = 0.1,
    subMenuCloseDelay = 0.1,
    forceSubMenuRender,
    defaultOpenKeys,
    openKeys,

    // Active
    activeKey,
    defaultActiveFirst,

    // Selection
    selectable = true,
    multiple = false,
    defaultSelectedKeys,
    selectedKeys,
    onSelect,
    onDeselect,

    // Level
    inlineIndent = navMenuDefaults.inlineIndent,

    // Motion
    motion,
    defaultMotions,

    // Popup
    triggerSubMenuAction = 'hover',
    builtinPlacements,

    // Icon
    itemIcon,
    expandIcon,
    overflowedIndicator = '...',

    // Function
    getPopupContainer,

    // Events
    onClick,
    onOpenChange,
    onKeyDown,

    // Deprecated
    openAnimation,
    openTransitionName,

    ...restProps
  } = props as LegacyMenuProps;

  const childList: React.ReactElement[] = parseChildren(children, EMPTY_LIST);
  const [mounted, setMounted] = React.useState(false);

  const containerRef = React.useRef<HTMLUListElement>(null);

  const uuid = useUUID(id);

  const isRtl = direction === 'rtl';

  // ========================= Warn =========================
  if (process.env.NODE_ENV !== 'production') {
    warning(
      !openAnimation && !openTransitionName,
      '`openAnimation` and `openTransitionName` is removed. Please use `motion` or `defaultMotion` instead.',
    );
  }

  // ========================= Open =========================
  const [mergedOpenKeys, setMergedOpenKeys] = useMergedState(defaultOpenKeys, {
    value: openKeys,
    postState: keys => keys || EMPTY_LIST,
  });

  // React 18 will merge mouse event which means we open key will not sync
  // ref: https://github.com/ant-design/ant-design/issues/38818
  const triggerOpenKeys = (keys: string[], forceFlush = false) => {
    function doUpdate() {
      setMergedOpenKeys(keys);
      onOpenChange?.(keys);
    }
    if (forceFlush) {
      flushSync(doUpdate);
    } else {
      doUpdate();
    }
  };

  // >>>>> Cache & Reset open keys when inlineCollapsed changed
  const [inlineCacheOpenKeys, setInlineCacheOpenKeys] = React.useState(mergedOpenKeys);

  const mountRef = React.useRef(false);

  // ========================= Mode =========================
  const [mergedMode, mergedInlineCollapsed] = React.useMemo<[RcMenuMode, boolean]>(() => {
    if ((mode === 'inline' || mode === 'vertical') && inlineCollapsed) {
      return ['vertical', inlineCollapsed];
    }
    return [mode, false];
  }, [mode, inlineCollapsed]);

  const isInlineMode = mergedMode === 'inline';

  const [internalMode, setInternalMode] = React.useState(mergedMode);
  const [internalInlineCollapsed, setInternalInlineCollapsed] = React.useState(mergedInlineCollapsed);

  React.useEffect(() => {
    setInternalMode(mergedMode);
    setInternalInlineCollapsed(mergedInlineCollapsed);

    if (!mountRef.current) {
      mountRef.current = true;
      return;
    }
    // Synchronously update MergedOpenKeys
    if (isInlineMode) {
      setMergedOpenKeys(inlineCacheOpenKeys);
    } else {
      // Trigger open event in case its in control
      triggerOpenKeys(EMPTY_LIST);
    }
  }, [mergedMode, mergedInlineCollapsed]);

  // ====================== Responsive ======================
  const [lastVisibleIndex, setLastVisibleIndex] = React.useState(0);
  const allVisible = lastVisibleIndex >= childList.length - 1 || internalMode !== 'horizontal' || disabledOverflow;

  // Cache
  React.useEffect(() => {
    if (isInlineMode) {
      setInlineCacheOpenKeys(mergedOpenKeys);
    }
  }, [mergedOpenKeys]);

  // ========================= Path =========================
  const {
    registerPath,
    unregisterPath,
    refreshOverflowKeys,

    isSubPathKey,
    getKeyPath,
    getKeys,
    getSubPathKeys,
  } = useKeyRecords();

  const registerPathContext = React.useMemo(() => ({ registerPath, unregisterPath }), [registerPath, unregisterPath]);

  const pathUserContext = React.useMemo(() => ({ isSubPathKey }), [isSubPathKey]);

  React.useEffect(() => {
    refreshOverflowKeys(
      allVisible ? EMPTY_LIST : childList.slice(lastVisibleIndex + 1).map(child => child.key as string),
    );
  }, [lastVisibleIndex, allVisible]);

  // ======================== Active ========================
  const [mergedActiveKey, setMergedActiveKey] = useMergedState<string | undefined>(
    activeKey || ((defaultActiveFirst && childList[0]?.key) as string),
    {
      value: activeKey,
    },
  );

  const onActive = useMemoCallback((key: string) => {
    setMergedActiveKey(key);
  });

  const onInactive = useMemoCallback(() => {
    setMergedActiveKey(undefined);
  });

  // ======================== Select ========================
  // >>>>> Select keys
  const [mergedSelectKeys, setMergedSelectKeys] = useMergedState(defaultSelectedKeys || [], {
    value: selectedKeys,

    // Legacy convert key to array
    postState: keys => {
      if (Array.isArray(keys)) {
        return keys;
      }

      if (keys === null || keys === undefined) {
        return EMPTY_LIST;
      }

      return [keys];
    },
  });

  // >>>>> Trigger select
  const triggerSelection = (info: RcMenuInfo) => {
    if (selectable) {
      // Insert or Remove
      const { key: targetKey } = info;
      const exist = mergedSelectKeys.includes(targetKey);
      let newSelectKeys: string[];

      if (multiple) {
        if (exist) {
          newSelectKeys = mergedSelectKeys.filter(key => key !== targetKey);
        } else {
          newSelectKeys = [...mergedSelectKeys, targetKey];
        }
      } else {
        newSelectKeys = [targetKey];
      }

      setMergedSelectKeys(newSelectKeys);

      // Trigger event
      const selectInfo: MenuSelectInfo = {
        ...info,
        selectedKeys: newSelectKeys,
      };

      if (exist) {
        onDeselect?.(selectInfo);
      } else {
        onSelect?.(selectInfo);
      }
    }

    // Whatever selectable, always close it
    if (!multiple && mergedOpenKeys?.length && internalMode !== 'inline') {
      triggerOpenKeys(EMPTY_LIST);
    }
  };

  // ========================= Open =========================
  /**
   * Click for item. SubMenu do not have selection status
   */
  const onInternalClick = useMemoCallback((info: RcMenuInfo) => {
    onClick?.(warnItemProp(info));
    triggerSelection(info);
  });

  const onInternalOpenChange = useMemoCallback((key: string, open: boolean) => {
    let newOpenKeys = mergedOpenKeys?.filter(k => k !== key) || [];

    if (open) {
      newOpenKeys.push(key);
    } else if (internalMode !== 'inline') {
      // We need find all related popup to close
      const subPathKeys = getSubPathKeys(key);
      newOpenKeys = newOpenKeys.filter(k => !subPathKeys.has(k));
    }

    if (!shallowEqual(mergedOpenKeys, newOpenKeys)) {
      triggerOpenKeys(newOpenKeys, true);
    }
  });

  const getInternalPopupContainer = useMemoCallback(getPopupContainer);

  // ==================== Accessibility =====================
  const triggerAccessibilityOpen = (key: string, open?: boolean) => {
    const nextOpen = open ?? !mergedOpenKeys?.includes(key);

    onInternalOpenChange(key, nextOpen);
  };

  const onInternalKeyDown = useAccessibility(
    internalMode,
    mergedActiveKey!,
    isRtl,
    uuid!,

    containerRef,
    getKeys,
    getKeyPath,

    setMergedActiveKey,
    triggerAccessibilityOpen,

    onKeyDown,
  );

  // ======================== Effect ========================
  React.useEffect(() => {
    setMounted(true);
  }, []);

  // ======================== Render ========================

  // >>>>> Children
  const wrappedChildList =
    internalMode !== 'horizontal' || disabledOverflow
      ? childList // Need wrap for overflow dropdown that do not response for open
      : childList.map((child, index) => (
          // Always wrap provider to avoid sub node re-mount
          <MenuContextProvider key={child.key} overflowDisabled={index > lastVisibleIndex}>
            {child}
          </MenuContextProvider>
        ));

  // >>>>> Container
  const container = (
    <Overflow
      id={id}
      ref={containerRef as any}
      prefixCls={`${prefixCls}-overflow`}
      component="ul"
      itemComponent={RcMenuItem}
      className={clsx(prefixCls, `${prefixCls}-root`, `${prefixCls}-${internalMode}`, className, {
        [`${prefixCls}-inline-collapsed`]: internalInlineCollapsed,
        [`${prefixCls}-rtl`]: isRtl,
      })}
      dir={direction}
      style={style}
      role="menu"
      tabIndex={tabIndex}
      data={wrappedChildList}
      renderRawItem={node => node}
      renderRawRest={omitItems => {
        // We use origin list since wrapped list use context to prevent open
        const len = omitItems.length;

        const originOmitItems = len ? childList.slice(-len) : null;

        return (
          <RcSubMenu
            eventKey={OVERFLOW_KEY}
            title={overflowedIndicator}
            disabled={allVisible}
            internalPopupClose={len === 0}
          >
            {originOmitItems}
          </RcSubMenu>
        );
      }}
      maxCount={internalMode !== 'horizontal' || disabledOverflow ? Overflow.INVALIDATE : Overflow.RESPONSIVE}
      ssr="full"
      data-menu-list
      onVisibleChange={newLastIndex => {
        setLastVisibleIndex(newLastIndex);
      }}
      onKeyDown={onInternalKeyDown}
      {...restProps}
    />
  );

  // >>>>> Render
  return (
    <IdContext.Provider value={uuid!}>
      <MenuContextProvider
        prefixCls={prefixCls}
        mode={internalMode}
        openKeys={mergedOpenKeys}
        rtl={isRtl}
        // Disabled
        disabled={disabled}
        // Motion
        motion={mounted ? motion : undefined}
        defaultMotions={mounted ? defaultMotions : undefined}
        // Active
        activeKey={mergedActiveKey}
        onActive={onActive}
        onInactive={onInactive}
        // Selection
        selectedKeys={mergedSelectKeys}
        // Level
        inlineIndent={inlineIndent}
        // Popup
        subMenuOpenDelay={subMenuOpenDelay}
        subMenuCloseDelay={subMenuCloseDelay}
        forceSubMenuRender={forceSubMenuRender}
        builtinPlacements={builtinPlacements}
        triggerSubMenuAction={triggerSubMenuAction}
        getPopupContainer={getInternalPopupContainer}
        // Icon
        itemIcon={itemIcon}
        expandIcon={expandIcon}
        // Events
        onItemClick={onInternalClick}
        onOpenChange={onInternalOpenChange}
      >
        <PathUserContext.Provider value={pathUserContext}>{container}</PathUserContext.Provider>

        {/* Measure menu keys. Add `display: none` to avoid some developer miss use the Menu */}
        <div style={{ display: 'none' }} aria-hidden>
          <PathRegisterContext.Provider value={registerPathContext}>{childList}</PathRegisterContext.Provider>
        </div>
      </MenuContextProvider>
    </IdContext.Provider>
  );
};

export default RcMenu;
