// @universe-design/react/src/menu/internal-menu/MenuItemGroup.tsx
import * as React from 'react';
import clsx from 'clsx';
import { parseChildren } from './utils/nodeUtil';
import { RcMenuContext } from './context/MenuContext';
import { useFullPath, useMeasure } from './context/PathContext';

export interface RcMenuItemGroupProps {
  className?: string;
  /**
   * 分组标题
   */
  title?: React.ReactNode;
  /**
   * 分组的菜单项
   */
  children?: React.ReactNode;

  /** @private Internal filled key. Do not set it directly */
  eventKey?: string;
}

const InternalMenuItemGroup = ({ className, title, eventKey, children, ...restProps }: RcMenuItemGroupProps) => {
  const { prefixCls } = React.useContext(RcMenuContext)!;

  const groupPrefixCls = `${prefixCls}-item-group`;

  return (
    <li {...restProps} onClick={e => e.stopPropagation()} className={clsx(groupPrefixCls, className)}>
      <div className={`${groupPrefixCls}-title`} title={typeof title === 'string' ? title : undefined}>
        {title}
      </div>
      <ul className={`${groupPrefixCls}-list`}>{children}</ul>
    </li>
  );
};

export default function RcMenuItemGroup({ children, ...props }: RcMenuItemGroupProps): React.ReactElement {
  const connectedKeyPath = useFullPath(props.eventKey);
  const childList: React.ReactElement[] = parseChildren(children, connectedKeyPath);

  const measure = useMeasure();
  if (measure) {
    return childList as any as React.ReactElement;
  }

  return <InternalMenuItemGroup {...props}>{childList}</InternalMenuItemGroup>;
}
