// @universe-design/react/es/shared/components/popper/imperative.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { Popper } from './Popper';
import { CrossRootConfigProvider } from '../../../config-provider';
import { isFunction } from '../../../util';
const ImperativePopperReference = /*#__PURE__*/React.forwardRef((props, ref) => {
  React.useImperativeHandle(ref, () => props.reference);
  return /*#__PURE__*/React.createElement(React.Fragment, null);
});
const ImperativePopper = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    defaultProps
  } = _ref;
  const [props, setProps] = React.useState(defaultProps);
  React.useImperativeHandle(ref, () => ({
    update: nextProps => {
      setProps(nextProps);
    }
  }), []);
  return /*#__PURE__*/React.createElement(Popper, props);
});
export class ImperativePopperController {
  constructor(_props, popper) {
    _defineProperty(this, "_popperManager", void 0);
    // 用做 ReactDOM render 的 container
    _defineProperty(this, "_container", void 0);
    _defineProperty(this, "_imperativePopperRef", /*#__PURE__*/React.createRef());
    _defineProperty(this, "_isAlive", false);
    _defineProperty(this, "update", props => {
      this._initOrUpdate(props);
    });
    _defineProperty(this, "destroy", () => {
      this._cleaup();
    });
    _defineProperty(this, "_initOrUpdate", props => {
      var _this$_imperativePopp;
      if (!this._isAlive) {
        // console.log('_init', this, props);
        this._isAlive = true;
        this._container = document.createElement('div');
        document.body.appendChild(this._container);
        ReactDOM.render(/*#__PURE__*/React.createElement(CrossRootConfigProvider, null, /*#__PURE__*/React.createElement(ImperativePopper, {
          ref: this._imperativePopperRef,
          defaultProps: this.props2PopperProps(props)
        })), this._container);
        this._popperManager.controllers.add(this);
        return;
      }
      // console.log('_update', this, props);
      (_this$_imperativePopp = this._imperativePopperRef.current) === null || _this$_imperativePopp === void 0 ? void 0 : _this$_imperativePopp.update(this.props2PopperProps(props));
    });
    _defineProperty(this, "_cleaup", () => {
      this._popperManager.controllers.delete(this);
      if (this._container) {
        ReactDOM.unmountComponentAtNode(this._container);
        document.body.removeChild(this._container);
      }
      this._isAlive = false;
    });
    this._popperManager = popper;
    this._initOrUpdate(_props);
  }
  props2PopperProps(props) {
    const {
      reference,
      destroyOnHide = true,
      onAfterClose,
      ...popperProps
    } = props;
    return {
      // defaultVisible: true,
      popupPortalContainer: false,
      // 默认在 container 原位渲染
      ...popperProps,
      destroyOnHide,
      onAfterClose: () => {
        isFunction(onAfterClose) && onAfterClose();
        if (destroyOnHide) {
          this._cleaup();
        }
      },
      children: /*#__PURE__*/React.createElement(ImperativePopperReference, {
        reference: reference
      }),
      domOnly: false // 为了让 virtual element 生效需要使用 domOnly false
    };
  }
}
class PopperManager {
  constructor() {
    _defineProperty(this, "controllers", new Set());
    _defineProperty(this, "create", props => new ImperativePopperController(props, this));
  }
}
export const popper = new PopperManager();
