// @universe-design/react/es/shared/components/popper/PopperProvider.js
import * as React from 'react';
import { useControlledState } from '@universe-design/react-utilities';
import composeRefs from '../../../util/composeRefs';
import { usePopperShared } from './usePopper';
import { usePopperActionShared } from './usePopperAction';
import { deprecatedWarning, __SCREENSHOT__, warning } from '../../../util';
import { isOverflowed as defaultIsOverflow } from '../../../util/dom';
import { useCssScope } from '../../../config-provider';
import { PopperContextProvider } from './PopperContext';
export const PopperProvider = /*#__PURE__*/React.forwardRef((props, ref) => {
  const {
    visible: visibleProp,
    checkOverflow = false,
    defaultVisible = false,
    children,
    trigger: triggerProp,
    showAction: showActionProp,
    hideAction: hideActionProp,
    mouseEnterDelay,
    mouseLeaveDelay: mouseLeaveDelayProp,
    onVisibleChange,
    udTriggerRef,
    disabled = false,
    prefixCls,
    modifiers,
    __screenshotAlwaysOpenEnabled = false
  } = props;

  /**
   * Deprecated Warning
   */
  process.env.NODE_ENV !== "production" ? warning(!udTriggerRef, deprecatedWarning('Popper', 'udTriggerRef', 'ref')) : void 0;
  process.env.NODE_ENV !== "production" ? warning(!modifiers, deprecatedWarning('Popper', 'modifiers', ['eventListeners', 'enableMaxHeight'])) : void 0;
  const rootPrefixCls = useCssScope('popper', prefixCls)('');

  /**
   * Popper 打开状态
   */
  const [visible, _setVisible] = useControlledState(defaultVisible, visibleProp, {
    onChange: onVisibleChange
  });

  /**
   * inline
   */
  const inlineElementMousePositionRef = React.useRef({
    x: 0,
    y: 0
  });

  /**
   * Popper 对齐
   */
  const {
    referenceRef,
    referenceElementRef,
    referenceElement,
    popperLatestValues
  } = usePopperShared();
  const checkVisible = React.useCallback(_visible => {
    if (_visible && checkOverflow) {
      const overflowChecker = typeof checkOverflow === 'function' ? checkOverflow : defaultIsOverflow;
      const child = referenceElementRef.current;
      if (!child || !overflowChecker(child)) {
        _setVisible(false);
        return;
      }
    }
    _setVisible(_visible);
  }, [checkOverflow, _setVisible, referenceElementRef]);

  // 保证在截图时一定能弹出
  __SCREENSHOT__ && __screenshotAlwaysOpenEnabled &&
  // eslint-disable-next-line react-hooks/rules-of-hooks
  React.useLayoutEffect(() => {
    checkVisible(true);
  });

  /**
   * Popper 触发方式
   */
  const {
    handleReferenceClick,
    handleContextMenu,
    handleMouseEnter,
    handleMouseLeave,
    handleKeyDown,
    handleReferenceFocus,
    setVisible,
    showAction,
    hideAction,
    alignPoint,
    mouseLeaveDelay,
    setVisibleDelay
  } = usePopperActionShared({
    trigger: triggerProp,
    showAction: showActionProp,
    hideAction: hideActionProp,
    mouseEnterDelay,
    mouseLeaveDelay: mouseLeaveDelayProp
  }, {
    visible,
    setVisible: checkVisible,
    referenceRef,
    referenceElementRef,
    inlineElementMousePositionRef
  });
  const mergedPopperRef = React.useMemo(() => composeRefs(udTriggerRef, ref), [ref, udTriggerRef]);
  React.useImperativeHandle(mergedPopperRef, () => ({
    visible,
    setVisible,
    forcePopupAlign: () => {
      var _popperLatestValues$u;
      return (_popperLatestValues$u = popperLatestValues.update) === null || _popperLatestValues$u === void 0 ? void 0 : _popperLatestValues$u.call(popperLatestValues);
    }
  }), [popperLatestValues, setVisible, visible]);
  return /*#__PURE__*/React.createElement(PopperContextProvider, {
    value: {
      showAction,
      hideAction,
      alignPoint,
      visible,
      handleReferenceClick,
      handleContextMenu,
      handleMouseEnter,
      handleMouseLeave,
      handleKeyDown,
      handleReferenceFocus,
      setVisible,
      setVisibleDelay,
      referenceRef,
      referenceElement,
      referenceElementRef,
      popperLatestValues,
      disabled,
      rootPrefixCls,
      mouseLeaveDelay,
      inlineElementMousePositionRef
    }
  }, children);
});
if (process.env.NODE_ENV !== "production") {
  PopperProvider.displayName = 'PopperProvider';
}
