// @universe-design/react/es/shared/components/popper/PopperPopupVisibleController.js
import React from 'react';
import { Transition } from 'react-transition-group';
import { useTransitionGroupNoopNodeRef } from '../transition/useAnimatedVisible';
import { defaults } from './types';
import { isFunction } from '../../../util';
import { usePopperContext } from './PopperContext';
export const PopperPopupVisibleController = props => {
  const {
    timeout = defaults.timeout,
    forceRender = false,
    destroyOnHide = true,
    onAfterClose,
    onAfterOpen,
    children
  } = props;
  const {
    visible,
    popperLatestValues
  } = usePopperContext();
  const noopRef = useTransitionGroupNoopNodeRef();
  return /*#__PURE__*/React.createElement(Transition, {
    nodeRef: noopRef,
    timeout: timeout,
    appear: true,
    in: visible,
    mountOnEnter: !forceRender,
    unmountOnExit: destroyOnHide,
    onEnter: () => {
      var _popperLatestValues$u;
      (_popperLatestValues$u = popperLatestValues.update) === null || _popperLatestValues$u === void 0 ? void 0 : _popperLatestValues$u.call(popperLatestValues);
    },
    onEntered: () => {
      isFunction(onAfterOpen) && onAfterOpen();
    },
    onExited: () => {
      var _popperLatestValues$r;
      isFunction(onAfterClose) && onAfterClose();
      (_popperLatestValues$r = popperLatestValues.resetPlacement) === null || _popperLatestValues$r === void 0 ? void 0 : _popperLatestValues$r.call(popperLatestValues);
    }
  }, children);
};
