// @universe-design/react/es/shared/components/popper/usePopperAction.js
import React from 'react';
import { useDelayCallback } from '../../../hooks/useDelayCallback';
import {
// getRelatedTarget,
isNonNullable, updateRef } from '../../../util';
import { KeyboardKey } from '@universe-design/react-utilities';
import { contains } from '../../../util/dom';
import { useComplementEvent } from '../../../hooks/useComplementEvent';
import { defaults } from '../trigger';
import { useEventCallback } from '../../../hooks/useEventCallback';
export const PopperActionEnum = {
  click: 'click',
  contextMenu: 'contextMenu',
  hover: 'hover',
  focus: 'focus',
  blur: 'blur',
  mouseEnter: 'mouseEnter',
  mouseLeave: 'mouseLeave'
};
const KEYDOWN_ESC_FLAG = Symbol.for('keydown.esc.flag');

// // 🐱 debug 神器
// document.addEventListener('focusin', function debugFocusin(ev) {
//   console.trace(
//     'focusin',
//     'target',
//     {
//       textContent: (ev.target as any).textContent,
//       tagName: (ev.target as any).tagName,
//     },
//     'relatedTarget',
//     {
//       textContent: (ev.relatedTarget as any)?.textContent,
//       tagName: (ev.relatedTarget as any)?.tagName,
//     }
//   );
// });
// document.addEventListener('focusout', function debugFocusout(ev) {
//   console.log(
//     'focusout',
//     'target',
//     {
//       textContent: (ev.target as any).textContent,
//       tagName: (ev.target as any).tagName,
//     },
//     'relatedTarget',
//     {
//       textContent: (ev.relatedTarget as any)?.textContent,
//       tagName: (ev.relatedTarget as any)?.tagName,
//     }
//   );
// });

// 对 rc-menu-item 的 event 放在 domEvent 里
const getEvent = ev => 'domEvent' in ev ? ev.domEvent : ev;
export function usePopperActionShared(props, context) {
  const {
    showAction: showActionProp,
    hideAction: hideActionProp,
    trigger: triggerProp,
    mouseEnterDelay = defaults.mouseEnterDelay,
    mouseLeaveDelay = defaults.mouseLeaveDelay,
    alignPoint: alignPointProp
  } = props;
  const {
    visible,
    setVisible,
    referenceRef,
    referenceElementRef,
    inlineElementMousePositionRef
  } = context;
  const [showAction, hideAction] = (() => {
    const actions = typeof triggerProp === 'string' ? [triggerProp] : triggerProp !== null && triggerProp !== void 0 ? triggerProp : [];
    const showActions = new Set(showActionProp !== null && showActionProp !== void 0 ? showActionProp : []);
    const hideActions = new Set(hideActionProp !== null && hideActionProp !== void 0 ? hideActionProp : []);
    for (const action of actions) {
      if (action === 'hover') {
        showActions.add('mouseEnter');
        hideActions.add('mouseLeave');
        continue;
      }
      if (action === 'click') {
        showActions.add('click');
        hideActions.add('click');
        continue;
      }
      if (action === 'contextMenu') {
        showActions.add('contextMenu');
        hideActions.add('click');
        continue;
      }
    }
    return [[...showActions].join('_'), [...hideActions].join('_')];
  })();
  const alignPoint = isNonNullable(alignPointProp) ? alignPointProp : showAction.includes(PopperActionEnum.contextMenu);
  const [setVisibleDelay, cleanupDelaySetVisible] = useDelayCallback(setVisible);
  const setVisibleSync = useEventCallback(function () {
    cleanupDelaySetVisible();
    return setVisible(...arguments);
  });
  const updateReferenceRefWhenAlignPoint = useEventCallback(ev => {
    if (alignPoint) {
      // contextMenu 滚动跟随，contextElement 作为 anchorElement
      const {
        currentTarget
      } = ev;
      const mouseRect = new DOMRect(ev.clientX, ev.clientY, 0, 0);
      const origDOMRect = currentTarget.getBoundingClientRect();
      updateRef(referenceRef, {
        getBoundingClientRect: () => {
          const newDOMRect = currentTarget.getBoundingClientRect();
          return new DOMRect(newDOMRect.x - origDOMRect.x + mouseRect.x, newDOMRect.y - origDOMRect.y + mouseRect.y, 0, 0);
        },
        contextElement: currentTarget
      });
    }
  });
  const handleReferenceClick = useEventCallback(function (_ev) {
    const ev = getEvent(_ev);
    if (ev.isDefaultPrevented()) {
      return;
    }
    if (visible && hideAction.includes(PopperActionEnum.click)) {
      return setVisibleSync(false);
    }
    if (!visible && showAction.includes(PopperActionEnum.click)) {
      updateReferenceRefWhenAlignPoint(ev); // 只有打开的时候需要更新 virtual element
      return setVisibleSync(true);
    }
  });
  const handleContextMenu = useEventCallback(function (_ev) {
    const ev = getEvent(_ev);
    if (!showAction.includes(PopperActionEnum.contextMenu) || ev.isDefaultPrevented()) {
      return;
    }
    // 阻止浏览器默认的 ContextMenu
    ev.preventDefault();
    updateReferenceRefWhenAlignPoint(ev);
    setVisibleSync(true);
  });
  const handleMouseEnter = useEventCallback(function (_ev) {
    const ev = getEvent(_ev);
    inlineElementMousePositionRef.current = {
      x: ev.clientX,
      y: ev.clientY
    };
    // alignPoint 的时候如果没有点击 referenceElementRef 默认是空的，这种情况下跳过展开逻辑
    if (ev.isDefaultPrevented() || !referenceElementRef.current) {
      return;
    }
    // reference bubble 并且没有 showAction mouseEnter 就是跳过打开
    if (ev.currentTarget === referenceElementRef.current && !showAction.includes(PopperActionEnum.mouseEnter) && !visible) {
      return;
    }
    setVisibleDelay([true], mouseEnterDelay);
  });
  const handleMouseLeave = useEventCallback(function (_ev) {
    const ev = getEvent(_ev);
    if (!hideAction.includes(PopperActionEnum.mouseLeave) || ev.isDefaultPrevented()) {
      return;
    }
    setVisibleDelay([false], mouseLeaveDelay);
  });
  const handleKeyDown = useEventCallback(function (_ev) {
    const ev = getEvent(_ev);
    if (ev.isDefaultPrevented() || KEYDOWN_ESC_FLAG in ev.nativeEvent) {
      return;
    }
    if (ev.key === KeyboardKey.Escape) {
      setVisibleSync(false);
      if (visible) {
        Object.assign(ev.nativeEvent, {
          [KEYDOWN_ESC_FLAG]: true
        });
      }
    }
  });

  // 用于判断 focus hover 事件的关联目标是否在 parent 节点内
  // const containsRelatedTarget = React.useCallback(
  //   (
  //     parent: Node | null | undefined,
  //     ev: React.FocusEvent<Element, Element>
  //   ) => {
  //     const relatedTarget = getRelatedTarget(ev);
  //     return contains(parent, relatedTarget as Node);
  //   },
  //   []
  // );

  const handleReferenceFocus = useEventCallback(function () {
    // TODO 由于取消 blur 关闭逻辑，所以 focus 打开逻辑暂时取消
    // const ev = getEvent(_ev);
    // if (
    //   // 这里通过 mouseEnter 而非 focus 去作为 showAction，主要是因为我们把 focus blur 作为 hover unhover 的 keyboard A11y 平替。
    //   !showAction.includes(PopperActionEnum.mouseEnter) ||
    //   containsRelatedTarget(floatingRef.current, ev)
    // ) {
    //   return;
    // }
    // setVisibleDelay([true], mouseEnterDelay);
  });
  return {
    handleReferenceClick,
    handleContextMenu,
    handleMouseEnter,
    handleMouseLeave,
    handleKeyDown,
    handleReferenceFocus,
    showAction,
    hideAction,
    alignPoint,
    setVisible: setVisibleSync,
    setVisibleDelay,
    mouseLeaveDelay
  };
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface

export function usePopperPopupAction(props, context) {
  const {
    visible,
    setVisibleDelay,
    floatingRef,
    referenceElementRef
  } = context;
  const portaledElementRef = React.useRef({
    mousedown: null
  });
  const recordPortal = (type, currentNode, source) => {
    if (!contains(currentNode, source)) {
      // 说明是 portal 传入的 click 事件，记录一下在 click outside 的时候用
      portaledElementRef.current[type] = source;
    }
  };
  const handlePopupFocus = useEventCallback(function () {
    // TODO 由于取消 blur 关闭逻辑，所以 focus 打开逻辑暂时取消
  });

  /**
   * Click outside
   */

  // 由于 useComplementEvent 要用 mousedown 作为关闭事件，所以必须要在 document 执行 mousedown 之前记录 portaled element，此处不能用 click
  const handlePopperMouseDownCapture = useEventCallback(function (_ev) {
    const ev = getEvent(_ev);
    recordPortal('mousedown', ev.currentTarget, ev.target);
  });

  // 这里 必须要用 mousedown 事件来保证多个 trigger 轮换的时候 return focus 不会错位。如果有 A 和 B 两个元素，开启 focuslock，A 打开的情况下点击 B，mousedown 的是时候会触发 focus 导致 A 的 focuslock 重新 lock 回 A，click 完成的时候 B focuslock 启动，这时候会认为是 A 的 trigger 是需要 returnTo 的元素。如果 mousedown 的时候就关闭 A 的 focuslock，则 B 的 trigger focus 可以保留
  useComplementEvent('mousedown', {
    listener: React.useCallback(function () {
      // 点击导致的 focusout 会在 micro task 的 returnFocus 和下一个 task 之间进行，所以关闭操作必须要用 task 异步
      setVisibleDelay([false], 0);
    }, [setVisibleDelay]),
    elements: React.useMemo(() => [floatingRef, referenceElementRef, () => portaledElementRef.current.mousedown], [floatingRef, referenceElementRef]),
    skipped: !visible
  });
  return {
    handlePopupFocus,
    handlePopperMouseDownCapture
  };
}
