// @universe-design/react/es/layout/Layout.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React, { forwardRef, useState } from 'react';
import clsx from 'clsx';
import { useCssScope } from '../config-provider/hooks';
import './style';
export const LayoutContext = /*#__PURE__*/React.createContext({
  siderControl: {
    addSider: () => null,
    removeSider: () => null
  }
});
const Layout = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    className,
    hasSider,
    children,
    ...restProps
  } = props;
  const [siders, setSiders] = useState([]);
  const cssScope = useCssScope('layout');
  const cls = clsx({
    [cssScope()]: true,
    [cssScope('--has-sider')]: !(hasSider === false) && (hasSider || siders.length > 0)
  }, className);
  return /*#__PURE__*/React.createElement(LayoutContext.Provider, {
    value: {
      siderControl: (() => ({
        addSider: id => {
          setSiders(prev => [...prev, id]);
        },
        removeSider: id => {
          setSiders(prev => prev.filter(currentId => currentId !== id));
        }
      }))()
    }
  }, /*#__PURE__*/React.createElement("section", _extends({
    ref: ref,
    className: cls
  }, restProps), children));
});
export default Layout;
