// @universe-design/react/es/layout/Sider.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React, { forwardRef, useCallback, useContext, useEffect, useRef, useState } from 'react';
import clsx from 'clsx';
import { useControlledState } from '@universe-design/react-utilities';
import { useEffectOnce } from '../hooks';
import { LayoutContext } from './Layout';
import { useCssScope } from '../config-provider/hooks';
import { isFunction } from '../util';
import SideExpandOutlined from '@universe-design/icons-react/SideExpandOutlined';
import SideFoldOutlined from '@universe-design/icons-react/SideFoldOutlined';
import DisordeListOutlined from '@universe-design/icons-react/DisordeListOutlined';
import './style';
const dimensionMaxMap = {
  xs: '479.98px',
  sm: '575.98px',
  md: '767.98px',
  lg: '991.98px',
  xl: '1199.98px',
  xxl: '1599.98px'
};
export const SiderContext = /*#__PURE__*/React.createContext(null);
if (process.env.NODE_ENV !== "production") {
  SiderContext.displayName = 'SiderContext';
}

/**
 * @public
 */

/**
 * @public
 * Layout.Sider 的属性和方法
 */

const generateId = (() => {
  let i = 0;
  return function () {
    let prefix = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
    i += 1;
    return `${prefix}${i}`;
  };
})();

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const isNumeric = value => !isNaN(parseFloat(value)) && isFinite(value);

// eslint-disable-next-line max-lines-per-function
const Sider = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    theme,
    style,
    className,
    width = 200,
    trigger,
    collapsed,
    collapsedWidth = 64,
    defaultCollapsed,
    collapsible,
    reverseArrow,
    zeroWidthTriggerStyle,
    breakpoint,
    children,
    onCollapse,
    onBreakpoint,
    ...othereProps
  } = props;
  const cssScope = useCssScope('layout__sider');
  const prefixCls = cssScope();
  const mqlRef = useRef();
  const {
    siderControl
  } = useContext(LayoutContext);
  const [below, setBelow] = useState(false);
  const [ignoreState, setIgnoreState] = useState(false);
  const [collapsedState, setCollapsedState] = useControlledState(Boolean(defaultCollapsed), collapsed);
  const setCollapsed = useCallback((collapsedValue, type) => {
    setCollapsedState(collapsedValue);
    isFunction(onCollapse) && onCollapse(collapsedValue, type);
  }, [onCollapse, setCollapsedState]);
  const responsiveHandler = useCallback(mql => {
    setBelow(mql.matches);
    isFunction(onBreakpoint) && onBreakpoint(mql.matches);
    if (!mql.matches) {
      setIgnoreState(false);
    }
    if (!ignoreState && collapsedState !== mql.matches) {
      setCollapsed(mql.matches, 'responsive');
    }
  }, [collapsedState, ignoreState, onBreakpoint, setCollapsed, setIgnoreState]);
  const toggle = () => {
    setCollapsed(!collapsedState, 'clickTrigger');
    setIgnoreState(true);
  };
  if (!mqlRef.current && typeof window !== 'undefined' && window.matchMedia && breakpoint) {
    mqlRef.current = window.matchMedia(`(max-width: ${dimensionMaxMap[breakpoint]})`);
    (mqlRef.current.addEventListener || mqlRef.current.addListener)('change', responsiveHandler);
    responsiveHandler(mqlRef.current);
  }
  useEffectOnce(() => {
    const uuid = generateId(cssScope());
    siderControl.addSider(uuid);
    return () => {
      siderControl.removeSider(uuid);
    };
  });
  useEffect(() => {
    let mql = null;
    if (!mql && typeof window !== 'undefined' && window.matchMedia && breakpoint) {
      mql = window.matchMedia(`(max-width: ${dimensionMaxMap[breakpoint]})`);
      if (mql.addEventListener) {
        mql.addEventListener('change', responsiveHandler);
      } else {
        mql.addListener(responsiveHandler);
      }
      responsiveHandler(mql);
    }
    return () => {
      if (mql) {
        if (mql.removeEventListener) {
          mql.removeEventListener('change', responsiveHandler);
        } else {
          mql.removeListener(responsiveHandler);
        }
        mql = null;
      }
    };
  }, [breakpoint, responsiveHandler]);
  const rawWidth = collapsedState ? collapsedWidth : width;
  const siderWidth = isNumeric(rawWidth) ? `${rawWidth}px` : String(rawWidth);
  const zeroWidthTrigger = parseFloat(String(collapsedWidth || 0)) === 0 ? /*#__PURE__*/React.createElement("span", {
    onClick: toggle,
    className: clsx(cssScope('--zero-width-trigger'), cssScope(`--zero-width-trigger-${reverseArrow ? 'right' : 'left'}`)),
    style: zeroWidthTriggerStyle
  }, trigger || /*#__PURE__*/React.createElement(DisordeListOutlined, null)) : null;
  const SiderStyle = {
    ...style,
    flex: `0 0 ${siderWidth}`,
    maxWidth: siderWidth,
    minWidth: siderWidth,
    width: siderWidth,
    '--collapsedWidth': typeof collapsedWidth === 'number' ? `${collapsedWidth}px` : collapsedWidth
  };
  const siderCls = clsx(className, prefixCls, {
    [cssScope(`__${theme}`)]: theme,
    [cssScope('--collapsed')]: Boolean(collapsedState),
    [cssScope('--has-trigger')]: collapsible && trigger !== null && !zeroWidthTrigger,
    [cssScope('--below')]: Boolean(below),
    [cssScope('--zero-width')]: parseFloat(siderWidth) === 0
  });
  const iconObj = {
    expanded: reverseArrow ? /*#__PURE__*/React.createElement(SideFoldOutlined, null) : /*#__PURE__*/React.createElement(SideExpandOutlined, null),
    collapsed: reverseArrow ? /*#__PURE__*/React.createElement(SideExpandOutlined, null) : /*#__PURE__*/React.createElement(SideFoldOutlined, null)
  };
  const status = collapsedState ? 'collapsed' : 'expanded';
  const defaultTrigger = iconObj[status];
  return /*#__PURE__*/React.createElement(SiderContext.Provider, {
    value: {
      siderCollapsed: collapsedState,
      collapsedWidth
    }
  }, /*#__PURE__*/React.createElement("aside", _extends({
    ref: ref,
    style: SiderStyle,
    className: siderCls
  }, othereProps), /*#__PURE__*/React.createElement("div", {
    className: cssScope('__children')
  }, children), (() => {
    if (collapsible || below && zeroWidthTrigger && trigger !== null) {
      return zeroWidthTrigger || /*#__PURE__*/React.createElement("div", {
        className: `${prefixCls}-trigger`,
        onClick: toggle,
        style: {
          width: siderWidth
        }
      }, trigger || defaultTrigger);
    }
    return null;
  })()));
});
export default Sider;
