// @universe-design/react/lib/shared/components/picker/Comp/PickerTrigger.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _util = require("../../../../util");
var _hooks = require("../../../../config-provider/hooks");
var _usePopperSidecar = require("../../../../config-provider/hooks/usePopperSidecar");
var _SlideUpDownTransition = require("../../transition/SlideUpDownTransition");
var _reactProvider = require("@universe-design/react-provider");
const PickerTrigger = function (props) {
  const {
    disabled,
    open,
    children,
    dropdownClassName,
    popupStyle,
    popupElement,
    setOpened,
    getPopupContainer
  } = props;
  const triggerElementEnable = Boolean((0, _util.isFunction)(children) || children) && !disabled;
  const cssScope = (0, _hooks.useCssScope)('picker');
  const prefixCls = cssScope();
  const Popper = (0, _usePopperSidecar.usePopperSidecar)();
  const showAction = triggerElementEnable ? ['click'] : [];
  const direction = (0, _hooks.useDirection)();
  const ctxGetPopupContainer = (0, _reactProvider.usePopupPortalContainer)();
  return /*#__PURE__*/_react.default.createElement(Popper, {
    action: undefined,
    showAction: showAction,
    hideAction: [],
    destroyOnHide: true,
    popupStyle: popupStyle,
    getPopupContainer: triggerNode => {
      const _getPopupContainer = getPopupContainer !== null && getPopupContainer !== void 0 ? getPopupContainer : ctxGetPopupContainer;
      return _getPopupContainer(triggerNode);
    },
    prefixCls: prefixCls,
    popupPlacement: direction === 'rtl' ? 'bottomRight' : 'bottomLeft',
    visible: open,
    popupClassName: (0, _clsx.default)(dropdownClassName, `${prefixCls}-dropdown`),
    onVisibleChange: setOpened,
    popup: popupElement,
    popupMotion: "slide-up",
    enableRestoreFocus: false,
    initialFocus: null,
    showArrow: false,
    arrowHasStroke: false,
    Transition: _SlideUpDownTransition.SlideUpDownTransition,
    enableMaxHeight: false,
    defaultMainAxisOffsetWithoutArrow: 4
  }, children);
};
var _default = exports.default = PickerTrigger;
