// @universe-design/react/src/date-picker/panels/base/Body.tsx
import clsx from 'clsx';
import type moment from 'moment';
import * as React from 'react';
import { useCssScope } from '../../../config-provider/hooks';
import Cell from '../../cell';
import type { ICellStatus } from '../../cell/Range.class';
import { useDateCalendarUIContext } from '../DateCalendar/DateCalendarUIContext';

interface IPanelBodyProps {
  rowNum: number;
  colNum: number;
  baseDate: moment.Moment;
  getCellDate: (date: moment.Moment, offset: number) => moment.Moment;
  getCellStatus: (date: moment.Moment) => ICellStatus;
  dateRender?: (currentDate: moment.Moment) => React.ReactNode;
  className?: string;
  cellClassName?: string;
  headerCells?: React.ReactNode;
  children?: React.ReactNode;
  autoHeight?: boolean;
}

function Body(props: IPanelBodyProps) {
  const {
    className,
    cellClassName,
    rowNum,
    colNum,
    baseDate,
    autoHeight,
    getCellDate = (date, _offset) => date,
    dateRender = () => null,
    getCellStatus,
    // headerCells,
    children,
  } = props;
  const prefixCls = useCssScope('picker')();
  const bodyPrefixCls = `${prefixCls}-panel-body`;
  const { onDateMouseEnter, onDateMouseLeave, onSelect, onContentMouseEnter, onContentMouseLeave } =
    useDateCalendarUIContext();
  const rows: React.ReactNode[] = [];

  for (let rowIndex = 0; rowIndex < rowNum; rowIndex++) {
    const isRowSelected = [...new Array(colNum)].some(
      (_, colIndex) => getCellStatus(getCellDate(baseDate, rowIndex * colNum + colIndex)).isSelected,
    );
    rows.push(
      <div className={clsx(`${bodyPrefixCls}-row`, isRowSelected && `${bodyPrefixCls}-row--selected`)} key={rowIndex}>
        {[...new Array(colNum)].map((_, colIndex) => {
          const offset = rowIndex * colNum + colIndex;
          const cellDate = getCellDate(baseDate, offset);
          const cellStatus = getCellStatus(cellDate);
          const { isDisabled } = cellStatus;

          const preCellDate = getCellDate(baseDate, offset - 1);
          const nextCellDate = getCellDate(baseDate, offset + 1);

          const prevCellStatus = getCellStatus(preCellDate);
          const nextCellStatus = getCellStatus(nextCellDate);

          return (
            <React.Fragment key={colIndex}>
              <Cell
                onClick={() => {
                  if (!isDisabled) {
                    onSelect(cellDate);
                  }
                }}
                onMouseEnter={() => {
                  if (!isDisabled) {
                    onDateMouseEnter(cellDate);
                  }
                }}
                onMouseLeave={() => {
                  if (!isDisabled) {
                    onDateMouseLeave(cellDate);
                  }
                }}
                prevCellStatus={prevCellStatus}
                nextCellStatus={nextCellStatus}
                className={cellClassName}
                colIndex={colIndex}
                colNum={colNum}
                {...cellStatus}
              >
                {dateRender(cellDate)}
              </Cell>
            </React.Fragment>
          );
        })}
      </div>,
    );
  }

  return (
    <div className={clsx(bodyPrefixCls, autoHeight && `${prefixCls}-auto-height`, className)}>
      {children}
      <div
        className={clsx(`${bodyPrefixCls}-content`, 'notranslate')}
        onMouseEnter={onContentMouseEnter}
        onMouseLeave={onContentMouseLeave}
      >
        {rows}
      </div>
    </div>
  );
}

export default Body;
