// @universe-design/react/src/date-picker/panels/base/utils.ts
import type moment from 'moment';
import type { MODE } from '../../interface';

export const YEAR_COUNT = 20;

export const getYearStart = (date: moment.Moment) => {
  const newDate = date.clone();
  return newDate.year(Math.floor(newDate.year() / YEAR_COUNT) * YEAR_COUNT);
};

export const getYearRange = (date: moment.Moment) => {
  const start = getYearStart(date).year();
  return [start, start + YEAR_COUNT - 1].join(' - ');
};

function generatorDateByMode(date: moment.Moment, mode: MODE, offset: number) {
  const newDate = date.clone();
  switch (mode) {
    case 'date':
      return newDate.add(offset, 'month');
    case 'month':
      return newDate.add(offset, 'year');
    case 'year':
      // return newDate.add(offset * YEAR_COUNT, 'year');
      return getYearStart(newDate).add(offset * YEAR_COUNT, 'year');
    default:
      return newDate;
  }
}

export const getNext = (date: moment.Moment, mode: MODE) => generatorDateByMode(date, mode, 1);

export const getPrev = (date: moment.Moment, mode: MODE) => generatorDateByMode(date, mode, -1);

export const getHigherLevelMode = (mode: MODE): MODE | null => {
  const levels = ['date', 'month', 'year'] as const;
  const index = levels.findIndex(m => m === mode);
  if (index < levels.length - 1) {
    return levels[index + 1];
  }
  return null;
};

export function isInRange(date: moment.Moment, range: [moment.Moment, moment.Moment], mode: MODE) {
  return date.isSameOrAfter(range[0], mode) && date.isSameOrBefore(range[1], mode);
}
