// @universe-design/react-utilities/lib-commonjs/components/checkbox/useCheckbox.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useCheckbox = void 0;
var _react = require("react");
var _clsx = _interopRequireDefault(require("clsx"));
var _useIsomorphicLayoutEffect = require("../../hooks/useIsomorphicLayoutEffect");
var _useControlledState = require("../../hooks/useControlledState");
/**
 * checkbox 组件的 hook
 * @public
 * @param props - checkbox 的 props
 * @param inputRef - input ref
 */
const useCheckbox = (props, inputRef) => {
  const {
    defaultChecked,
    checked,
    name,
    style,
    className,
    error,
    disabled,
    indeterminate,
    onChange,
    classPrefix
  } = props;
  const [checkedState, setCheckedState] = (0, _useControlledState.useControlledState)(Boolean(defaultChecked), Boolean(checked), {
    isControlledOverride: 'checked' in props
  });
  const syncIndeterminate = (0, _react.useCallback)(() => {
    if (indeterminate !== undefined && inputRef.current) {
      inputRef.current.indeterminate = indeterminate;
    }
  }, [indeterminate, inputRef]);
  (0, _useIsomorphicLayoutEffect.useIsomorphicLayoutEffect)(() => {
    syncIndeterminate();
  }, [syncIndeterminate]);
  return {
    checked: checkedState,
    indeterminate,
    inputProps: {
      name,
      disabled,
      type: 'checkbox',
      checked: checkedState,
      onChange: e => {
        syncIndeterminate();
        setCheckedState(e.target.checked, () => {
          onChange === null || onChange === void 0 ? void 0 : onChange(e.target.checked, e);
        });
      },
      role: 'checkbox',
      ['aria-checked']: function () {
        var _inputRef$current;
        if (indeterminate || (_inputRef$current = inputRef.current) !== null && _inputRef$current !== void 0 && _inputRef$current.indeterminate) {
          return 'mixed';
        }
        return checkedState;
      }(),
      className: (0, _clsx.default)(`${classPrefix}__input`, disabled && `${classPrefix}__input-disabled`)
    },
    wrapperProps: {
      style,
      className: (0, _clsx.default)(`${classPrefix}__wrapper`, {
        [(0, _clsx.default)(`${classPrefix}--error`)]: error,
        [(0, _clsx.default)(`${classPrefix}--disabled`)]: disabled
      }, className)
    },
    descriptionProps: {
      className: (0, _clsx.default)(`${classPrefix}__label-content`, disabled && `${classPrefix}__label-content-disabled`)
    }
  };
};
exports.useCheckbox = useCheckbox;
