// @universe-design/react/src/dialog/dialog/index.tsx
/* eslint-disable max-nested-callbacks */
import stylex from '@universe-design/stylex';
import clsx from 'clsx';
import * as React from 'react';
import { RemoveScrollBar } from 'react-remove-scroll-bar';
import { CSSTransition, Transition } from 'react-transition-group';
import { unstable_useConfigReducedMotion as useConfigReducedMotion } from '../../config-provider/ConfigReducedMotion';
import { createGlobalConfig } from '@universe-design/react-provider';
import { useCssScope } from '../../config-provider/hooks';
import { useIsomorphicLayoutEffect as useLayoutEffect } from '../../hooks';
import Portal from '../../portal';
import { FocusLock } from '../../shared';
import { useAnimatedVisible } from '../../shared/components/transition/useAnimatedVisible';
import type { DefaultExtraSizeType, DefaultSizeType, FocusableElement } from '../../util';
import { KeyCode, devWarning, getSizeAbbr, getWindow, isNonNullable, parseLength, warning } from '../../util';
import composeRefs from '../../util/composeRefs';
import { forwardHTMLRef } from '../../util/forwardRef';
import type { ThunkElement } from '../../util/getThunkElement';
import { isClient } from '@universe-design/react-utilities';

import '../style';

/**
 * @deprecated
 */
export interface DialogLayoutModeProps {
  /**
   * 使用的对话框布局模式。
   * - `legacy`，使用版本小于 2.18.0 的旧版对话框布局
   * - `2.18.0`，使用版本 2.18.0 的对话框布局
   */
  mode: 'legacy' | '2.18.0';
  children?: React.ReactNode;
}

const {
  ConfigProvider,
  useConfig: useDialogLayoutMode,
  setGlobal,
} = /* #__PURE__*/ createGlobalConfig<DialogLayoutModeProps['mode']>('dialog-layout-mode', '2.18.0');

export { useDialogLayoutMode };

/**
 * @deprecated
 * 为了更好支持设计规范，我们在 `2.18.0` 修改了对话框的布局方式，由于修改布局方式，对于之前覆盖过对话框样式的用户可能产生 breaking change。默认情况下，我们会使用 `2.18.0` 新布局，如果用户侧存量代码较多，验证成本较高，可以使用该组件修改子树中的对话框布局模式为 `legacy`。
 *
 * 建议渐进式验证并替换为新版 layout，本组件将会在下个大版本中移除
 */
export const DialogLayoutMode = Object.assign(
  function DialogLayoutMode({ mode, children }: DialogLayoutModeProps) {
    return <ConfigProvider value={mode}>{children}</ConfigProvider>;
  },
  {
    setGlobal,
  },
);

export enum DialogCancelType {
  'CancelButton',
  'CloseIcon',
  'Esc',
  'Mask',
}

function useDialogCssScope() {
  return useCssScope('dialog');
}

export interface DialogExoticProps {
  /**
   * 对话框是否可见
   */
  visible?: boolean;
  /**
   * 指定对话框挂载的 HTML 节点
   * @defaultValue `document.body`
   */
  getContainer?: string | HTMLElement | (() => HTMLElement) | false | null;
  /**
   * 垂直居中展示，默认推荐使用垂直居中布局。如果使用垂直居中布局，`style.top` 属性会失效
   * @defaultValue true
   */
  centered?: boolean;
  /**
   * 对话框内容宽度，设置后内置的尺寸预设值失效，相当于 `minWidth = maxWidth = width`，定宽尺寸（仅支持 `number` 类型的数字、`string` 类型的 `%` 和 `px`）
   */
  width?: number | string;
  /**
   * 对话框最大宽度，优先级高于 `width`，可以覆盖设置最小宽度
   */
  minWidth?: number | string;
  /**
   * 对话框最大宽度，优先级高于 `width`，可以覆盖设置最大宽度
   */
  maxWidth?: number | string;
  /**
   * 设置对话框的 `z-index`
   * @defaultValue 1000
   */
  zIndex?: number;
  overlay?: boolean;
  overlayClosable: boolean;
  /**
   * 是否支持键盘 Esc 关闭
   * @defaultValue true
   */
  keyboardClosable?: boolean;
  overlayStyle?: React.CSSProperties;
  contentStyle?: React.CSSProperties;
  overlayClassName?: string;
  contentClassName?: string;
  wrapperClassName?: string;
  /**
   * 是否开启动画
   * @defaultValue true
   */
  enableAnimation?: boolean;
  overlayTransitionClassName?: string;
  contentTransitionClassName?: string;
  /**
   * 关闭时销毁对话框里的子元素
   * @defaultValue false
   */
  destroyOnClose?: boolean;
  /**
   * 首次出现前是否强制渲染子元素
   * @defaultValue false
   */
  forceRender?: boolean;
  onClose?: (
    e: React.MouseEvent<HTMLElement, MouseEvent> | React.KeyboardEvent | undefined,
    payload: { cancelType: DialogCancelType },
  ) => void;
  onAfterClose?: () => void;
  onAfterOpen?: () => void;
  /**
   * 对话框尺寸，小尺寸、中尺寸为固定值，大尺寸和超大尺寸为响应式，[详情](/component/pc/dialog)请见
   */
  size?: DefaultSizeType | Extract<DefaultExtraSizeType, 'extra-large'>;
  /**
   * 安全边距，`top` 为竖直上边安全边距，`bottom` 为竖直下边安全边距，`hmin` 为水平最小边距，`hmax` 为水平最大边距（未达到 `maxWidth` 时的最大边距）
   * @defaultValue \{ top: 32, bottom: 32, hmax: 180, hmin: 32 \}
   */
  safeMargin?: {
    top?: number;
    bottom?: number;
    hmax?: number;
    hmin?: number;
  };
  /**
   * 进入 dialog 时的初始 focus 的元素, 默认为 content
   */
  initialFocusRef?: ThunkElement<FocusableElement>;
  /**
   * dialog 关闭后恢复 focus 的元素
   */
  finalFocusRef?: ThunkElement<FocusableElement>;
  /**
   * 开启🤔关闭 focusLock
   * @defaultValue false
   */
  enableFocusLock?: boolean;
  /**
   * 关闭对话框后是否恢复 focus 到打开它的元素
   * @defaultValue false
   */
  enableRestoreFocus?: boolean;
}

const defaultSafeMarginHmin = 32;

const defaults: {
  getContainer: NonNullable<DialogExoticProps['getContainer']>;
  safeMargin: Required<NonNullable<DialogExoticProps['safeMargin']>>;
} = {
  getContainer: isClient() ? () => document.body : false,
  safeMargin: {
    top: 32,
    bottom: 32,
    hmax: 180,
    hmin: defaultSafeMarginHmin,
  },
};

const defaultSizeCfg = {
  small: {
    minWidth: 420,
    maxWidth: 420,
  },
  medium: {
    minWidth: 600,
    maxWidth: 600,
  },
  large: {
    minWidth: 420,
    maxWidth: 840,
  },
  'extra-large': {
    minWidth: 420,
    maxWidth: 1080,
  },
};

export interface DialogProps extends DialogExoticProps, React.HTMLAttributes<HTMLDivElement> {}

const DialogContent = forwardHTMLRef<
  Required<Pick<DialogExoticProps, 'centered' | 'safeMargin' | 'size'>> &
    Pick<DialogExoticProps, 'width' | 'minWidth' | 'maxWidth'> & {
      in: boolean;
    },
  'div'
>(
  (
    {
      centered,
      safeMargin,
      size,
      width: widthProp,
      minWidth: minWidthProp,
      maxWidth: maxWidthProp,
      style: styleProp = {},
      in: inProp,
      ...restProps
    },
    ref,
  ) => {
    // safeMargin
    const {
      top: safeMarginTop = defaults.safeMargin.top,
      bottom: safeMarginBottom = defaults.safeMargin.bottom,
      hmax: safeMarginHMax = defaults.safeMargin.hmax,
      hmin: safeMarginHMin = defaults.safeMargin.hmin,
    } = safeMargin;
    const contentRef = React.useRef<HTMLDivElement>(null);
    const mergedContentRef = React.useMemo(() => composeRefs(ref, contentRef), [ref]);

    /**
     * 调整 dialog content 的 top left width height，用于自适应安全边距和最大最小宽度高度。
     * **注意**: 这里 height 相关的实现和技术设计中的描述有一些出入，如果要获取准确的 minHeight 和 maxHeight 需要修改容器的 maxHeight 值再获取 scrollHeight，但是这个过程有两个问题：
     * 1. 会引起 force reflow，会有性能问题 https://gist.github.com/paulirish/5d52fb081b3570c81e3a
     * 2. scrollTop 会失效，需要 reset 后代节点的 scrollTop
     *
     * 为了避免不必要的性能损耗，这里尽可能通过 style 状态和父容器状态进行 nextStyle 的推算，给节点做 nextStyle 赋值，从而走浏览器自己的 batch update 流程。有必要的话可以把这个方法暴露给用户。
     */
    const adjustContentInset = React.useCallback(
      function adjustContentInset() {
        const contentEl = contentRef.current;
        if (!contentEl) {
          return;
        }

        if (!contentEl.offsetParent) {
          return;
        }

        const computedStyle = getComputedStyle(contentEl);

        // minWidth & maxWidth
        // 优先级 minWidthProp > style.minWidth > computedStyle.minWidth > widthProp > style.width > computedStyle.width > size config
        const minWidth = (() => {
          if (isNonNullable(minWidthProp)) {
            return minWidthProp;
          }
          if (isNonNullable(styleProp.minWidth)) {
            return styleProp.minWidth;
          }
          if (computedStyle.minWidth !== '0px' && !contentEl.style.minWidth) {
            // css minWidth
            return computedStyle.minWidth;
          }
        })();

        const maxWidth = (() => {
          if (isNonNullable(maxWidthProp)) {
            return maxWidthProp;
          }
          if (isNonNullable(styleProp.maxWidth)) {
            return styleProp.maxWidth;
          }
          if (computedStyle.maxWidth !== 'none' && !contentEl.style.maxWidth) {
            return computedStyle.maxWidth;
          }
        })();

        const win = getWindow(contentEl);
        // safari 下 scrollHeight 不准所以用 getBoundingClientRect
        const containerRect = contentEl.parentElement?.getBoundingClientRect();
        const containerWidth = containerRect?.width ?? win.innerWidth;
        const containerHeight = containerRect?.height ?? win.innerHeight;
        const _width = (() => {
          if (isNonNullable(widthProp)) {
            return widthProp;
          }
          if (isNonNullable(styleProp.width)) {
            return styleProp.width;
          }
        })();
        const minWidthNum = parseLength(minWidth ?? _width ?? defaultSizeCfg[size].minWidth, containerWidth);
        const maxWidthNum = parseLength(maxWidth ?? _width ?? defaultSizeCfg[size].maxWidth, containerWidth);

        /**
         * 处理宽度
         */
        const { width, marginH } = (function handleWidth() {
          // width 算出来可能是一个非整数，我们不做处理，不然可能会产生 width 抖动导致 reflow
          if (containerWidth < minWidthNum + safeMarginHMin * 2) {
            return {
              width: `${minWidthNum}px`,
              marginH: `${safeMarginHMin}px`,
            };
          }
          if (containerWidth < minWidthNum + safeMarginHMax * 2) {
            const margin = (containerWidth - minWidthNum) / 2;
            return {
              width: `${containerWidth - margin * 2}px`,
              marginH: `${margin}px`,
            };
          }
          if (containerWidth < maxWidthNum + safeMarginHMax * 2) {
            return {
              width: `${containerWidth - safeMarginHMax * 2}px`,
              marginH: `${safeMarginHMax}px`,
            };
          }
          const margin = (containerWidth - maxWidthNum) / 2;
          return {
            width: `${containerWidth - margin * 2}px`,
            marginH: `${margin}px`,
          };
        })();

        /**
         * 处理高度
         */
        const { height, marginTop } = (function handleHeight() {
          const _top = (() => {
            if (isNonNullable(styleProp.top)) {
              return styleProp.top;
            }
            if (computedStyle.top !== '0px' && !contentEl.style.top) {
              return computedStyle.top;
            }
            return centered ? 0 : 100;
          })();
          const contentTop = parseLength(_top, containerHeight);
          if (centered) {
            const margin = Math.max(safeMarginTop, safeMarginBottom);
            return {
              height: `${containerHeight - margin * 2}px`,
              marginTop: `${margin}px`,
            };
          }
          if (contentTop < safeMarginTop) {
            return {
              height: `${containerHeight - safeMarginTop - safeMarginBottom}px`,
              marginTop: `${safeMarginTop}px`,
            };
          }
          return {
            height: `${containerHeight - contentTop - safeMarginBottom}px`,
            marginTop: `${contentTop}px`,
          };
        })();

        // 用 marginRight 为了在 wrapper 水平滚动的时候右侧可以撑开，但是很不幸只有最新版的 chrome 是这个行为，其他版本 margin 都是不能占位的。
        // 不用 padding 有两个原因，一是 padding 会导致 mask 无法点击到，需要特殊处理，二是会跟之前的实现有些出入，可能导致用户挂掉
        // TODO marginBottom 在竖直滚动的时候也无法撑开
        contentEl.style.width = width;
        contentEl.style.marginLeft = marginH;
        contentEl.style.marginRight = marginH;
        contentEl.style.height = height;
        contentEl.style.top = marginTop;
      },
      [
        widthProp,
        styleProp.width,
        styleProp.minWidth,
        styleProp.maxWidth,
        styleProp.top,
        minWidthProp,
        size,
        maxWidthProp,
        safeMarginHMin,
        safeMarginHMax,
        centered,
        safeMarginTop,
        safeMarginBottom,
      ],
    );

    useLayoutEffect(() => {
      if (inProp) {
        adjustContentInset();
      }
    }, [adjustContentInset, inProp]);

    React.useEffect(() => {
      const win = getWindow(contentRef.current);
      const cb = () => adjustContentInset();
      win.addEventListener('resize', cb);
      return () => {
        win.removeEventListener('resize', cb);
      };
    });

    return <div ref={mergedContentRef} style={styleProp} {...restProps} />;
  },
);

const DialogContentLegacy = forwardHTMLRef<Required<Pick<DialogExoticProps, 'centered' | 'safeMargin'>>, 'div'>(
  ({ centered, safeMargin, ...restProps }, ref) => {
    warning(
      false,
      devWarning(
        'Dialog',
        // eslint-disable-next-line max-len
        "You are using `Dialog` in legacy mode, legacy mode has been deprecated and will be removed in the next major release. The following properties cannot be used: `minWidth` `maxWidth` `safeMargin.hmin` `safeMargin.hmax` `maxHeight` `bodyHeightResponsiveMode` in legacy mode. Please remove <DialogLayoutMode mode='legacy' /> or wrap your <Dialog /> with <DialogLayoutMode mode='2.18.0' /> for better support for design specs.",
      ),
    );
    const { top: safeMarginTop = defaults.safeMargin.top, bottom: safeMarginBottom = defaults.safeMargin.bottom } =
      safeMargin;
    const contentRef = React.useRef<HTMLDivElement>(null);
    React.useImperativeHandle<HTMLDivElement | null, HTMLDivElement | null>(ref, () => contentRef.current, []);
    useLayoutEffect(() => {
      // 矫正 content top 不超过安全边距
      if (contentRef.current) {
        let contentTop = parseFloat(getComputedStyle(contentRef.current).top || '0');
        if (!centered && contentTop < safeMarginTop) {
          contentTop = safeMarginTop;
          contentRef.current.style.top = `${safeMarginTop}px`;
        } else if (centered && contentTop > 0) {
          // top 属性失效
          contentTop = 0;
          contentRef.current.style.top = '0px';
        }
        contentRef.current.style.maxHeight = `calc(100vh - ${
          centered ? safeMarginTop + safeMarginBottom : safeMarginBottom + contentTop
        }px)`;
      }
    });
    return <div ref={contentRef} {...restProps} />;
  },
);

// eslint-disable-next-line max-lines-per-function
export const Dialog = React.forwardRef<HTMLDivElement, DialogProps>((props, ref) => {
  const {
    visible = false,
    overlay = true,
    overlayClosable,
    keyboardClosable = true,
    contentStyle,
    overlayStyle,
    overlayClassName,
    contentClassName,
    wrapperClassName,
    enableAnimation: enableAnimationProp,
    overlayTransitionClassName,
    contentTransitionClassName,
    destroyOnClose,
    children,
    width,
    zIndex,
    centered = true,
    getContainer = defaults.getContainer,
    onClose,
    onAfterClose,
    onAfterOpen,
    size = 'medium',
    safeMargin = defaults.safeMargin,
    className,
    initialFocusRef,
    finalFocusRef,
    enableFocusLock = false,
    enableRestoreFocus = false,
    forceRender,
    minWidth,
    maxWidth,
    ...restProps
  } = props;

  const dialogCssScope = useDialogCssScope();
  const layoutMode = useDialogLayoutMode();
  const enableAnimation = !useConfigReducedMotion(
    isNonNullable(enableAnimationProp) ? !enableAnimationProp : undefined,
  );

  const wrapperRef = React.useRef<HTMLDivElement>(null);
  const contentRef = React.useRef<HTMLDivElement>(null);
  const contentClickRef = React.useRef<boolean>();
  const contentTimeoutRef = React.useRef<number>();

  // 添加 esc 监听
  const onKeyDown = React.useCallback<React.KeyboardEventHandler<HTMLDivElement>>(
    e => {
      if (keyboardClosable) {
        const { keyCode } = e;
        /* istanbul ignore else  */
        if (keyCode === KeyCode.ESC) {
          onClose && onClose(e, { cancelType: DialogCancelType.Esc });
        }
      }
    },
    [keyboardClosable, onClose],
  );

  const handleOverlayClick = (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
    if (!overlayClosable) {
      return;
    }
    if (contentClickRef.current) {
      contentClickRef.current = false;
    } else if (wrapperRef.current === e.target) {
      onClose && onClose(e, { cancelType: DialogCancelType.Mask });
    }
  };

  const onContentMouseDown: React.MouseEventHandler = () => {
    window.clearTimeout(contentTimeoutRef.current);
    contentClickRef.current = true;
  };

  const onContentMouseUp: React.MouseEventHandler = () => {
    contentTimeoutRef.current = window.setTimeout(() => {
      contentClickRef.current = false;
    });
  };

  const { transitionGroupProps, transitionGroupPropsWithoutListener, styleDisplay, noopRef } = useAnimatedVisible({
    destroyOnClose,
    forceRender,
    visible,
    onAfterClose,
    onAfterOpen,
  });

  const timeout = React.useMemo(
    () => ({
      // eslint-disable-next-line @typescript-eslint/no-magic-numbers
      enter: enableAnimation ? 400 : 0,
      // eslint-disable-next-line @typescript-eslint/no-magic-numbers
      exit: enableAnimation ? 300 : 0,
    }),
    [enableAnimation],
  );

  const maskRef = React.useRef<HTMLDivElement>(null);
  const mergedInitialFocusRef = initialFocusRef === null ? null : initialFocusRef ?? wrapperRef;

  const renderOverlay = () => {
    if (!overlay) {
      return null;
    }
    const cls = clsx(dialogCssScope('__mask'), overlayClassName);

    let animationCls = '';
    if (enableAnimation) {
      animationCls = overlayTransitionClassName || dialogCssScope('__mask-animation');
    }
    const mergeStyle: React.CSSProperties = {
      ...overlayStyle,
      ...styleDisplay,
    };
    if (zIndex || zIndex === 0) {
      mergeStyle.zIndex = zIndex;
    }
    return (
      <CSSTransition
        timeout={timeout}
        appear
        classNames={animationCls}
        nodeRef={maskRef}
        {...transitionGroupPropsWithoutListener}
      >
        <div ref={maskRef} style={mergeStyle} className={cls} />
      </CSSTransition>
    );
  };

  const usedSize = getSizeAbbr(size);
  const renderContent = () => {
    const wrapCls = clsx(
      dialogCssScope('__wrap'),
      wrapperClassName,
      {
        [dialogCssScope('--centered')]: centered,
      },
      stylex(
        layoutMode === 'legacy'
          ? [
              centered &&
                stylex.create({
                  textAlign: 'center',
                  // 固定宽度小于屏幕时会换行，这里阻止一下
                  whiteSpace: 'nowrap',
                  '::before': {
                    display: 'inline-block',
                    width: 0,
                    height: '100%',
                    verticalAlign: 'middle',
                    boxSizing: 'border-box',
                    content: "''",
                  },
                }),
            ]
          : [
              centered &&
                stylex.create({
                  // 固定宽度小于屏幕时会换行，这里阻止一下
                  whiteSpace: 'nowrap',
                }),
            ],
      ),
    );
    const cls = clsx(
      dialogCssScope('__content'),
      contentClassName,
      {
        [dialogCssScope(`__content-${usedSize}`)]: usedSize,
      },
      stylex(
        layoutMode === 'legacy'
          ? [
              stylex.create({
                position: 'relative',
                boxSizing: 'border-box',
                top: '100px',
                margin: '0 auto',
                // 重置 white-space
                whiteSpace: 'normal',
              }),
              centered &&
                stylex.create({
                  top: 0,
                  display: 'inline-flex',
                  textAlign: 'left',
                  verticalAlign: 'middle',
                }),
              usedSize &&
                {
                  sm: stylex.create({
                    width: 420,
                  }),
                  md: stylex.create({
                    width: 600,
                    minHeight: 128,
                  }),
                  lg: stylex.create({
                    minWidth: 420,
                    width: `calc(100vw - ${defaultSafeMarginHmin}px * 2)`,
                    maxWidth: 840,
                  }),
                  xl: stylex.create({
                    minWidth: 420,
                    width: `calc(100vw - ${defaultSafeMarginHmin}px * 2)`,
                    maxWidth: 1080,
                  }),
                }[usedSize],
            ]
          : [
              stylex.create({
                position: 'relative',
                // 重置 white-space
                whiteSpace: 'normal',
                display: 'flex',
                alignItems: 'flex-start',
                pointerEvents: 'none',
              }),
            ],
      ),
    );
    const wrapStyle: React.CSSProperties = {
      ...styleDisplay,
    };
    if (zIndex || zIndex === 0) {
      wrapStyle.zIndex = zIndex;
    }
    let animationCls = '';
    if (enableAnimation) {
      animationCls = contentTransitionClassName || dialogCssScope('__content-animation');
    }
    return (
      <CSSTransition
        timeout={timeout}
        appear
        classNames={animationCls}
        nodeRef={contentRef} // 这里改为 contentRef 因为在用 ScrollArea 的时候和之前用 div 的时候行为不同，关闭 dialog 会有偶现的闪动
        {...transitionGroupPropsWithoutListener}
      >
        <FocusLock
          isDisabled={!visible}
          initialFocusRef={mergedInitialFocusRef}
          restoreFocus={enableRestoreFocus}
          finalFocusRef={finalFocusRef}
          enableFocusLock={enableFocusLock}
        >
          <div
            className={wrapCls}
            style={wrapStyle}
            tabIndex={-1}
            onClick={handleOverlayClick}
            onKeyDown={onKeyDown}
            ref={wrapperRef}
          >
            {layoutMode === 'legacy' ? (
              <DialogContentLegacy
                style={{
                  ...contentStyle,
                  width,
                }}
                className={cls}
                onMouseDown={onContentMouseDown}
                onMouseUp={onContentMouseUp}
                safeMargin={safeMargin}
                centered={centered}
                role="dialog"
                aria-modal="true"
                ref={contentRef}
              >
                {children}
              </DialogContentLegacy>
            ) : (
              <DialogContent
                style={contentStyle}
                className={cls}
                onMouseDown={onContentMouseDown}
                onMouseUp={onContentMouseUp}
                safeMargin={safeMargin}
                centered={centered}
                role="dialog"
                aria-modal="true"
                ref={contentRef}
                size={size}
                width={width}
                minWidth={minWidth}
                maxWidth={maxWidth}
                in={transitionGroupPropsWithoutListener.in}
              >
                {children}
              </DialogContent>
            )}
          </div>
        </FocusLock>
      </CSSTransition>
    );
  };

  const renderPortal = () => {
    const dialogRootNode = (
      <div ref={ref} className={clsx(dialogCssScope('__root'), className)} {...restProps}>
        {visible && <RemoveScrollBar noRelative gapMode="margin" />}
        {renderOverlay()}
        {renderContent()}
      </div>
    );
    if (getContainer === false) {
      return dialogRootNode;
    }
    return <Portal container={getContainer}>{dialogRootNode}</Portal>;
  };

  return (
    <Transition timeout={timeout} appear nodeRef={noopRef} {...transitionGroupProps}>
      {renderPortal()}
    </Transition>
  );
});
