// @byted-service/bytedtrace/lib/span/consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ERROR_MESSAGE_KEY = exports.DEFAULT_SAMPLING_TRACE_OPTIONS = exports.CUSTOM_SPAN_LOG_TYPE = exports.CLIENT_SPAN_RPC_LOG_TYPE = exports.SERVER_SPAN_RPC_LOG_TYPE = exports.CUSTOM_SPAN_LOG_LOCATION = exports.ACCESS_LOG_LOCATION = exports.CALL_LOG_LOCATION = exports.CLIENT_SPAN_METRICS_TIMER_NAME = exports.CLIENT_SPAN_METRICS_COUNTER_NAME = exports.SERVICE_SPAN_METRICS_TIMER_NAME = exports.SERVICE_SPAN_METRICS_COUNTER_NAME = exports.SAMPLING_TRACE_ROOT_TYPE_TAG = exports.SPAN_UN_FINISHED = exports.SPAN_FINISHED = exports.DEFAULT_CLUSTER = void 0;
const tslib_1 = require("tslib");
const gConsts = tslib_1.__importStar(require("../consts"));
const types = tslib_1.__importStar(require("./types"));
exports.DEFAULT_CLUSTER = 'default';
exports.SPAN_FINISHED = 1;
exports.SPAN_UN_FINISHED = 0;
exports.SAMPLING_TRACE_ROOT_TYPE_TAG = '_sampling_trace_root_type';
exports.SERVICE_SPAN_METRICS_COUNTER_NAME = `span.${gConsts.SERVER_SPAN_TYPE}.rate`;
exports.SERVICE_SPAN_METRICS_TIMER_NAME = `span.${gConsts.SERVER_SPAN_TYPE}.latency.us`;
exports.CLIENT_SPAN_METRICS_COUNTER_NAME = `span.${gConsts.CLIENT_SPAN_TYPE}.rate`;
exports.CLIENT_SPAN_METRICS_TIMER_NAME = `span.${gConsts.CLIENT_SPAN_TYPE}.latency.us`;
exports.CALL_LOG_LOCATION = 'call.js:0';
exports.ACCESS_LOG_LOCATION = 'access.js:0';
exports.CUSTOM_SPAN_LOG_LOCATION = 'custom_span.js:0';
exports.SERVER_SPAN_RPC_LOG_TYPE = 'rpcAccess';
exports.CLIENT_SPAN_RPC_LOG_TYPE = 'rpcCall';
exports.CUSTOM_SPAN_LOG_TYPE = 'customSpan';
exports.DEFAULT_SAMPLING_TRACE_OPTIONS = {
    samplingTraceType: types.SamplingTraceType.postTraceType,
    rootSpanContext: undefined,
};
exports.ERROR_MESSAGE_KEY = '_err_msg';
