// @byted-service/bytedtrace/lib/span/handler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpanHandler = void 0;
const tslib_1 = require("tslib");
const gStatus = tslib_1.__importStar(require("../status"));
const types = tslib_1.__importStar(require("./types"));
const consts = tslib_1.__importStar(require("./consts"));
const utils_1 = require("../utils");
class SpanHandler {
    constructor() {
        this.spanId = '0';
        this.parentSpanId = '0';
        this.isRootSpan = false;
        this.spanRefType = 0;
        this.name = '';
        /**
         * 微秒
         */
        this.startTime = 0;
        /**
         * 微秒
         */
        this.duration = 0;
        this.tags = [];
        this.emitRemoteAddrInMetrics = false;
        this.postTrace = 0;
        // 避免重复执行finish
        this.isFinished = 0;
    }
    /**
     * 判断传入参数是不是 noop Span。
     * @param span
     * @returns
     */
    static isNoopSpan(span) {
        return span ? span.noop() : true;
    }
    /**
     * 设置状态码，例如在 client span 调用下游(通过 http client)返回状态码 200。
     * @param statusCode
     * @returns
     */
    setStatusCode(statusCode) {
        this.reservedParam.statusCode = statusCode;
        this.checkConditionSample();
        return this;
    }
    /**
     * 获取 span 类型
     * @returns
     */
    getSpanType() {
        return this.spanType;
    }
    /**
     * 获取 status code
     * @returns
     */
    getStatusCode() {
        return this.reservedParam.statusCode;
    }
    /**
     * 获取是否是 root span
     * @returns
     */
    getIsRootSpan() {
        return this.isRootSpan;
    }
    /**
     * 设置 span name
     * @param name
     * @returns
     */
    setSpanName(name) {
        this.name = name;
        return this;
    }
    /**
     * 设置是否有发生错误
     * @param isError
     * @returns
     */
    setIsError(isError) {
        this.reservedParam.isError = isError;
        this.checkConditionSample();
        return this;
    }
    /**
     * 获取 restrictedKey 对应在 baggage 中的 value 值
     * @param restrictedKey
     * @returns
     */
    baggageItem(restrictedKey) {
        if (!this.baggage && this.transaction.baggage) {
            return this.transaction.baggage.get(restrictedKey);
        }
        return this.baggage.get(restrictedKey);
    }
    /**
     * 获取 span name
     * @returns
     */
    getName() {
        return this.name;
    }
    /**
     * 获取是否发生错误的状态值
     * @returns
     */
    getIsError() {
        return this.reservedParam.isError;
    }
    /**
     * 设置条件过滤器
     * @param conditionFilters
     */
    setConditionFilters(conditionFilters) {
        this.conditionFilters = conditionFilters;
    }
    /**
     * 获取条件过滤器
     * @returns
     */
    getConditionFilters() {
        return this.conditionFilters;
    }
    /**
     * 设置业务状码
     * @param businessStatusCode
     * @returns
     */
    setBusinessStatusCode(businessStatusCode) {
        this.reservedParam.businessStatusCode = businessStatusCode;
        this.checkConditionSample();
        return this;
    }
    /**
     * 获取业务状态码
     * @returns
     */
    getBusinessStatusCode() {
        return this.reservedParam.businessStatusCode;
    }
    /**
     * 设置发送的数据长度，例如 client span 通过 http 请求的方式发送给下游的数据的长度(单位：字节)。
     * @param sendSize
     * @returns
     */
    setSendSize(sendSize) {
        this.reservedParam.sendSize = sendSize;
        return this;
    }
    /**
     * 设置接收的数据长度，例如 server span 接收到上游通过 http 请求的方式传送过来的数据的长度(单位：字节)。
     * @param recvSize
     * @returns
     */
    setRecvSize(recvSize) {
        this.reservedParam.recvSize = recvSize;
        return this;
    }
    /**
     * service type 主要标识数据协议，http thrift redis mysql 等。
     * @param serviceType
     * @returns
     */
    setServiceType(serviceType) {
        this.reservedParam.serviceType = serviceType;
        return this;
    }
    /**
     * 设置组件，lib 或框架名，例如 gulu。
     * @param component
     * @returns
     */
    setComponent(component) {
        this.reservedParam.component = component;
        return this;
    }
    /**
     * 设置上游的 P.S.M。
     * @param fromService
     * @returns
     */
    setFromService(fromService) {
        this.reservedParam.remoteService = fromService;
        return this;
    }
    /**
     * 设置上游的具体方法名。
     * @param fromMethod
     * @returns
     */
    setFromMethod(fromMethod) {
        this.reservedParam.remoteMethod = fromMethod;
        return this;
    }
    /**
     * 设置上游所在的集群名称。
     * @param fromCluster
     * @returns
     */
    setFromCluster(fromCluster) {
        this.reservedParam.remoteCluster = fromCluster;
        return this;
    }
    /**
     * 设置上游的 IP 或域名地址。
     * @param addr
     * @returns
     */
    setFromAddr(addr) {
        this.reservedParam.remoteAddrStr = addr;
        return this;
    }
    /**
     * 设置上游的 dc 名称。
     * @param fromDc
     * @returns
     */
    setFromDc(fromDc) {
        this.reservedParam.remoteDc = fromDc;
        return this;
    }
    /**
     * 设置被调用下游的方法名称。
     * @param toMethod
     * @returns
     */
    setToMethod(toMethod) {
        this.reservedParam.remoteMethod = toMethod;
        return this;
    }
    /**
     * 设置被调用下游的集群名称。
     * @param toCluster
     * @returns
     */
    setToCluster(toCluster) {
        this.reservedParam.remoteCluster = toCluster;
        return this;
    }
    /**
     * 设置被调用下游的 IP 或者域名地址，例如 127.0.0.1。
     * @param addr
     * @returns
     */
    setToAddr(addr) {
        this.reservedParam.remoteAddrStr = addr;
        return this;
    }
    /**
     * 设置被调用下游 dc 名称，例如 lf。
     * @param toDc
     * @returns
     */
    setToDc(toDc) {
        this.reservedParam.remoteDc = toDc;
        return this;
    }
    /**
     * 设置被调用下游的 P.S.M。
     * @param service
     * @returns
     */
    setToService(service) {
        this.reservedParam.remoteService = service;
        return this;
    }
    /**
     * 设置采样类型
     * @param opt
     * @returns
     */
    samplingTrace(opt) {
        this._samplingTrace(opt);
        return this;
    }
    /**
     * 判断是否允许向上游传播链路信息，如果调用 isBackwardPropagate 返回 true，需要将 spanContext 注入到 response 中传播到上游
     * @returns
     */
    isBackwardPropagate() {
        return this.transaction ? this.transaction.isPostTrace() : false;
    }
    /**
     * 判断是否允许向下游传播链路信息，如果调用 isForwardPropagate 返回 true，需要将 span 的 spanContext 注入到 request 中传播到下游。
     * @returns
     */
    isForwardPropagate() {
        return this.transaction ? this.transaction.isForwardPropagate() : false;
    }
    /**
     * 是否上报
     * @returns
     */
    isReported() {
        return this.transaction ? this.transaction.shouldReported() : false;
    }
    /**
     * 获取 baggage，这里不是返回 baggage 而是创建了一个新的
     * @returns
     */
    getBaggage() {
        return new Map(this.baggage);
    }
    /**
     * 获取 log id
     * @returns
     */
    getLogId() {
        const txn = this.transaction;
        return txn ? txn.logId : '';
    }
    /**
     * 获取 txn 的 method
     * @returns
     */
    getRootSpanMethod() {
        const txn = this.transaction;
        return txn ? txn.method : '';
    }
    /**
     * 设置额外需要透传的数据，在 inject 的时候会注入到 carrier 中，相同 key 后面的会覆盖前面的。
     * @param key
     * @param value
     * @returns
     */
    setBaggageItem(key, value) {
        if (!this.baggage) {
            const txn = this.transaction;
            if (txn && txn.baggage) {
                this.baggage = new Map(txn.baggage);
            }
            else {
                this.baggage = new Map();
            }
        }
        this.baggage.set(key, value);
        return this;
    }
    /**
     * 设置多个 tag
     * @param tagKVs
     * @returns
     */
    setTags(tagKVs) {
        if (this.getBreaker().getMaxTagKVCountInObj() -
            Object.keys(tagKVs).length >
            0) {
            const keys = Object.keys(tagKVs);
            for (let i = 0; i < keys.length; i++) {
                const key = keys[i];
                const val = tagKVs[key];
                this.tags.push([key, val]);
            }
        }
        return this;
    }
    /**
     * 获取 breaker
     * @returns
     */
    getBreaker() {
        return this.tracer.options.breaker;
    }
    /**
     * 设置父 span id
     * @param id
     */
    setParentSpanId(id) {
        this.parentSpanId = id;
    }
    /**
     * 设置后置采样，以下两个场景需要调用此方法进行后置采样：
     * 当前 span 没有被采样，但是下游进行了后置采样(第二种场景)，那么上游从 response 解析得到 span context 后需要设置后置采样，调用 setPostTrace，参数为解析得到 span context，可参考模拟上下游调用场景demo。https://code.byted.org/nodejs/byted-service/blob/master/packages/bytedtrace/examples/demo.ts
     * 当前 span 没有被采样，但是发生了一些研发人员感兴趣的事情(例如 catch 到了 error)，那么可以调用 setPostTrace 进行后置采样，不需要要传参数.
     * @param ctx
     */
    setPostTrace(ctx) {
        this._samplingTrace({
            samplingTraceType: types.SamplingTraceType.postTraceType,
            rootSpanContext: ctx,
        });
    }
    /**
     * 设置结束时间
     * @param t
     */
    setFinishTime(t) {
        this.duration = t - this.startTime;
        this.checkConditionSample();
    }
    /**
     * 设置单个 tag。
     * @param key
     * @param value
     * @returns
     */
    setTag(key, value) {
        if (this.tags.length <
            this.tracer.options.breaker.getMaxTagKVCountInObj()) {
            this.tags.push([key, value]);
        }
        else {
            /* istanbul ignore next */
            this.tracer.options.statusCounter.recordWarningWithRateLimit(false, gStatus.ErrSpanTagOverflow, 'tagKv in span exceed cap!');
        }
        return this;
    }
    /**
     * 设置错误信息。例如调用下游发生错误， client span 可以记录错误信息。
     * @param errorMsg
     * @returns
     */
    setErrorMessage(errorMsg) {
        this.setTag(consts.ERROR_MESSAGE_KEY, errorMsg);
        return this;
    }
    /**
     * 获取从 span 启动到当前的时间点一共消耗的时间。
     * @returns
     */
    getDurationCost() {
        if (!this.duration) {
            return (0, utils_1.nowInUs)() - this.startTime;
        }
        return this.duration;
    }
    /**
     * 设置 consumer group
     * @param consumerGroup
     */
    setConsumerGroup(consumerGroup) {
        this.extInfo.consumerGroup = consumerGroup;
    }
    /**
     * 设置生产者服务名(P.S.M)
     * @param producerService
     */
    setProducerService(producerService) {
        this.extInfo.producerService = producerService;
    }
    /**
     * 设置生产者集群名
     * @param producerCluster
     */
    setProducerCluster(producerCluster) {
        this.extInfo.producerCluster = producerCluster;
    }
    /**
     * 设置生产者地址
     * @param producerAddr
     */
    setProducerAddr(producerAddr) {
        this.extInfo.producerAddr = producerAddr;
    }
    /**
     * 设置生产者 dc
     * @param producerDc
     */
    setProducerDc(producerDc) {
        this.extInfo.producerDc = producerDc;
    }
    /**
     * 设置生产时间 单位为 ms
     * @param produceTs
     */
    setProduceTs(produceTs) {
        this.extInfo.produceTs = produceTs;
        if (!this.extInfo.mqLag) {
            this.extInfo.mqLag = this.startTime - produceTs;
        }
    }
    /**
     * 设置 mq 分区信息
     * @param mqPartition
     */
    setMqPartition(mqPartition) {
        this.extInfo.mqPartition = mqPartition;
    }
    /**
     * 设置堆积时间 单位 ms
     * @param mqLag
     */
    setMqLag(mqLag) {
        this.extInfo.mqLag = mqLag;
    }
    checkConditionSample() {
        const conditionFilters = this.getConditionFilters();
        if (conditionFilters) {
            conditionFilters.sample(this);
        }
    }
}
exports.SpanHandler = SpanHandler;
