// @byted-service/bytedtrace/lib/span/noop.js
"use strict";
/* tslint:disable max-classes-per-file */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NoopSpan = exports.defaultNoopSpanContext = exports.NoopSpanContext = void 0;
const handler_1 = require("./handler");
const tracerId_1 = require("../tracer/tracerId");
const extInfo_1 = require("./extInfo");
/* istanbul ignore next */
class NoopSpanContext {
    constructor() {
        this.baggage = new Map();
        this.traceID = new tracerId_1.TraceId(0, 0);
        this.flags = 0;
        this.spanID = '0';
        this.parentID = '0';
        // For example: dyeing
        this.debugInfo = '';
    }
    noop() {
        return true;
    }
    toString() {
        return '';
    }
    toStringW3C() {
        return '';
    }
    getBaggageItem() {
        return '';
    }
    isDebug() {
        return false;
    }
    isValid() {
        return false;
    }
}
exports.NoopSpanContext = NoopSpanContext;
exports.defaultNoopSpanContext = new NoopSpanContext();
/* istanbul ignore next */
class NoopSpan extends handler_1.SpanHandler {
    constructor(tracer) {
        super();
        this.reservedParam = {};
        this.tags = [];
        this.events = [];
        this.metricPbs = [];
        this.tracer = tracer;
        this.extInfo = new extInfo_1.ExtInfo();
    }
    init() {
        // do nothing
    }
    getDurationCost() {
        return 0;
    }
    synStartOpts() {
        // do nothing
    }
    reportMetrics() {
        // do nothing
    }
    emitSpanLog() {
        // do nothing
    }
    getBaggage() {
        return new Map();
    }
    isPostTrace() {
        return false;
    }
    getContext() {
        return exports.defaultNoopSpanContext;
    }
    samplingTrace() {
        return this;
    }
    _samplingTrace() {
        // do nothing
    }
    noop() {
        return true;
    }
    setPostTrace() {
        // do nothing
    }
    addEvents() {
        // do nothing
    }
    addEvent() {
        // do nothing
    }
    emitMetricsTimer() {
        // do nothing
    }
    emitMetricsRateCounter() {
        // do nothing
    }
    emitMetricsStore() {
        // do nothing
    }
    emitMetricsCounter() {
        // do nothing
    }
    emitMetricsMeter() {
        // do nothing
    }
    saveMetrics() {
        // do nothing
    }
    addMetrics() {
        // do nothing
    }
    isSampled() {
        return false;
    }
    setBaggageItem() {
        return this;
    }
    baggageItem() {
        return '';
    }
    isSampledLocal() {
        return false;
    }
    finish() {
        // do nothing
    }
    addLogAsEvent() {
        // do nothing
    }
    setFinishTime() {
        // do nothing
    }
    setTag() {
        return this;
    }
    resetReservedMetricsParam() {
        // do nothing
    }
    enableMetrics() {
        // do nothing
    }
}
exports.NoopSpan = NoopSpan;
