// @byted-service/bytedtrace/lib/span/span.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Span = void 0;
const tslib_1 = require("tslib");
const gTypes = tslib_1.__importStar(require("../types"));
const gConsts = tslib_1.__importStar(require("../consts"));
const gStatus = tslib_1.__importStar(require("../status"));
const gUtils = tslib_1.__importStar(require("../utils"));
const bytedtracePb = tslib_1.__importStar(require("../gen/bytedtrace_pb"));
const types = tslib_1.__importStar(require("./types"));
const consts = tslib_1.__importStar(require("./consts"));
const spanContext_1 = require("./spanContext");
const metricPb_1 = require("../metricPb");
const handler_1 = require("./handler");
const batchSpanInfo_1 = require("./batchSpanInfo");
const status_1 = require("../status");
const consumerSpanInfo_1 = require("./consumerSpanInfo");
class Span extends handler_1.SpanHandler {
    constructor(tracer) {
        super();
        this.tracer = tracer;
        this.tags = [];
        this.reservedParam = {
            statusCode: 0,
            businessStatusCode: 0,
            isError: false,
            sendSize: 0,
            recvSize: 0,
        };
        this.events = [];
        this.metricPbs = [];
    }
    /**
     * 判断是否是 ServerSpan
     * @returns
     */
    isServerSpan() {
        return this.spanType === gConsts.SERVER_SPAN_TYPE;
    }
    /**
     *  判断是否是 ClientSpan
     * @returns
     */
    isClientSpan() {
        return this.spanType === gConsts.CLIENT_SPAN_TYPE;
    }
    /**
     * 是否是 debug trace 类型
     * @returns
     */
    isDebug() {
        return this.transaction.isDebugTrace();
    }
    /**
     * 重置完成状态
     */
    resetFinish() {
        this.isFinished = consts.SPAN_UN_FINISHED;
    }
    /**
     * 重置 posttrace
     */
    resetPostTrace() {
        this.postTrace = 0;
    }
    /**
     * 判断是否是 posttrace
     * @returns
     */
    isPostTrace() {
        return this.transaction.isPostTrace();
    }
    /**
     * 给当前 span 添加多个 TraceEvent 可参考,BytedTracer.addEvents,
     * https://code.byted.org/nodejs/byted-service/blob/master/packages/bytedtrace/docs/API.md#addeventsevents-traceevent-span-span
     * @param events
     * @returns
     */
    addEvents(events) {
        if (events.length <= 0) {
            /* istanbul ignore next */
            return;
        }
        this.tracer._dealEvents(events, this);
        if (this.isSampledLocal()) {
            const appendNum = Math.min(events.length, this.tracer.options.breaker.getMaxEventCountInSpan() -
                this.events.length);
            if (appendNum < events.length) {
                this.tracer.options.statusCounter.recordWarningWithRateLimit(false, gStatus.ErrEventNumInSpanOverflow, 'events number excceed cap in span!');
            }
            const startTime = gUtils.nowInUs();
            for (let i = 0; i < appendNum; i++) {
                const event = events[i];
                if (event) {
                    if (!event.getStartTime()) {
                        event.setTimestamp(startTime);
                    }
                    this.events.push(event);
                }
            }
        }
    }
    /**
     * 给当前 span 添加 TraceEvent 可参考 BytedTracer.addEvents,
     * https://code.byted.org/nodejs/byted-service/blob/master/packages/bytedtrace/docs/API.md#addeventsevents-traceevent-span-span
     * @param event
     */
    addEvent(event) {
        this.addEvents([event]);
    }
    /**
     * 判断是否被采样
     * @returns
     */
    isSampled() {
        return this.transaction.isSampled();
    }
    /**
     * 判断是否是本地采样
     * @returns
     */
    isSampledLocal() {
        return this.tracer.allReported() || this.isSampled();
    }
    /**
     * 如果 span 被标为采样调用 finish 会将 span 数据上报到 argos。
     * @returns
     */
    finish() {
        if (!this.applyFinish()) {
            return;
        }
        if (!this.duration) {
            this.duration = gUtils.nowInUs() - this.startTime;
            this.checkConditionSample();
        }
        if (this.extInfo) {
            this.extInfo.appendExtraSpanTags(this.tags);
        }
        this.transaction.reportSpan(this);
    }
    /**
     * 返回 span 对应的 span context。
     * @returns
     */
    getContext() {
        const { transaction: txn } = this;
        let baggage = undefined;
        if (!this.baggage) {
            /* istanbul ignore next */
            baggage = txn.baggage ? new Map(txn.baggage) : new Map();
        }
        else {
            baggage = new Map(this.baggage);
        }
        const ctx = new spanContext_1.SpanContext();
        ctx.traceID = txn.traceID;
        ctx.spanID = this.spanId;
        ctx.parentID = this.parentSpanId;
        ctx.flags = txn.getFlags();
        ctx.baggage = baggage;
        ctx.debugInfo = txn.debugInfo;
        return ctx;
    }
    /**
     * timer 类型打点，首先会调用 BytedTracer 的同名方法 BytedTracer.emitMetricsTimer，
     * 即使用 @byted-service/metrics 打一次点，之后也会上传一份打点数据到 argos 平台，其他参数可参考 BytedTracer.emitMetricsTimer。
     * @param metricsName
     * @param value
     * @param tagKvs
     */
    emitMetricsTimer(metricsName, value, tagKvs) {
        this.tracer.emitMetricsTimer(metricsName, value, tagKvs);
        if (this.shouldReported()) {
            this.saveMetrics(gConsts.TIMER_TYPE, metricsName, value, tagKvs);
        }
    }
    /**
     * rate counter 类型打点，可参考 emitMetricsTimer。
     * @param metricsName
     * @param value
     * @param tagKvs
     */
    emitMetricsRateCounter(metricsName, value, tagKvs) {
        this.tracer.emitMetricsRateCounter(metricsName, value, tagKvs);
        if (this.shouldReported()) {
            this.saveMetrics(gConsts.RATE_COUNTER_TYPE, metricsName, value, tagKvs);
        }
    }
    /**
     * store 类型打点，可参考 emitMetricsTimer。
     * @param metricsName
     * @param value
     * @param tagKvs
     */
    emitMetricsStore(metricsName, value, tagKvs) {
        this.tracer.emitMetricsStore(metricsName, value, tagKvs);
        if (this.shouldReported()) {
            this.saveMetrics(gConsts.STORE_TYPE, metricsName, value, tagKvs);
        }
    }
    /**
     * counter 类型打点，可参考 emitMetricsTimer。
     * @param metricsName
     * @param value
     * @param tagKvs
     */
    emitMetricsCounter(metricsName, value, tagKvs) {
        this.tracer.emitMetricsCounter(metricsName, value, tagKvs);
        if (this.shouldReported()) {
            this.saveMetrics(gConsts.COUNTER_TYPE, metricsName, value, tagKvs);
        }
    }
    /**
     * meter 类型打点，可参考 emitMetricsTimer。
     * @param metricsName
     * @param value
     * @param tagKvs
     */
    emitMetricsMeter(metricsName, value, tagKvs) {
        this.tracer.emitMetricsMeter(metricsName, value, tagKvs);
        if (this.shouldReported()) {
            this.saveMetrics(gConsts.METER_TYPE, metricsName, value, tagKvs);
        }
    }
    /**
     * noopSpan 返回 false，其他类型的 span 返回 true。
     * @returns
     */
    noop() {
        return false;
    }
    _init(options) {
        this.name = options.name;
        this.spanType = options.spanType;
        this.spanRefType = options.spanRefType;
        this.transaction = options.txn;
        this.isRootSpan = options.isRootSpan;
        if (options.baggage) {
            this.baggage = options.baggage;
        }
    }
    _reportMetricsAndLog() {
        this.reportMetrics();
        this.emitSpanLog();
    }
    _synStartOpts(startOps = {}) {
        if (!startOps.startTime) {
            this.startTime = gUtils.nowInUs();
        }
        else {
            this.startTime = startOps.startTime;
        }
        this.extSpanModel =
            startOps.extSpanModel || gConsts.SpanModel.BaseSpanModel;
        switch (startOps.extSpanModel) {
            case gConsts.SpanModel.BatchSpanModel: {
                this.extInfo = new batchSpanInfo_1.BatchSpanInfo(startOps.disableBatchSpanTotalMetrics);
                break;
            }
            case gConsts.SpanModel.ConsumerSpanModel: {
                if (this.spanType !== gTypes.SpanType.server) {
                    this.tracer.options.statusCounter.recordWarningWithRateLimit(true, status_1.ErrMqConsumerSpanTypeInvalid, 'AsMqConsumerSpan must be used with server span');
                    return;
                }
                this.extInfo = new consumerSpanInfo_1.ConsumerSpanInfo();
                break;
            }
        }
        this.reference = startOps.reference || [];
        this.emitMetricsOpt = startOps.emitMetrics;
        this.emitSpanLogOpt = startOps.emitSpanLog;
        this.emitRemoteAddrInMetrics =
            startOps.emitRemoteAddrInMetrics || false;
    }
    _tranToServerSpanProto() {
        const breaker = this.getBreaker();
        const { reservedParam } = this;
        const spanPb = new bytedtracePb.ServerSpanProto();
        spanPb.setTagsList(breaker.buildTags(this.tags) || []);
        spanPb.setFromAddr(reservedParam.remoteAddrStr);
        spanPb.setMethod(this.name);
        spanPb.setRefType(this.spanRefType);
        spanPb.setReferencesList(this.buildSpanRefs());
        spanPb.setSpanId(this.spanId);
        spanPb.setParentId(this.parentSpanId);
        spanPb.setEventsList(this.genEventPbs());
        spanPb.setMetricsList(this.genMetricsProto());
        spanPb.setDuration(this.duration);
        spanPb.setStartTime(this.startTime);
        spanPb.setIsError(reservedParam.isError);
        spanPb.setBusinessStatusCode(reservedParam.businessStatusCode);
        spanPb.setStatusCode(reservedParam.statusCode);
        spanPb.setFromService(reservedParam.remoteService);
        spanPb.setFromMethod(reservedParam.remoteMethod);
        spanPb.setFromCluster(reservedParam.remoteCluster);
        spanPb.setFromDc(reservedParam.remoteDc);
        spanPb.setSendSize(reservedParam.sendSize);
        spanPb.setRecvSize(reservedParam.recvSize);
        spanPb.setServiceType(reservedParam.serviceType);
        spanPb.setComponent(reservedParam.component);
        return spanPb;
    }
    _tranToClientSpanProto() {
        const breaker = this.getBreaker();
        const { reservedParam } = this;
        const spanPb = new bytedtracePb.ClientSpanProto();
        spanPb.setTagsList(breaker.buildTags(this.tags) || []);
        spanPb.setToAddr(reservedParam.remoteAddrStr);
        spanPb.setRefType(this.spanRefType);
        spanPb.setReferencesList(this.buildSpanRefs());
        spanPb.setSpanId(this.spanId);
        spanPb.setParentId(this.parentSpanId);
        spanPb.setEventsList(this.genEventPbs());
        spanPb.setMetricsList(this.genMetricsProto());
        spanPb.setDuration(this.duration);
        spanPb.setStartTime(this.startTime);
        spanPb.setIsError(reservedParam.isError);
        spanPb.setBusinessStatusCode(reservedParam.businessStatusCode);
        spanPb.setStatusCode(reservedParam.statusCode);
        spanPb.setToService(reservedParam.remoteService);
        spanPb.setToMethod(reservedParam.remoteMethod);
        spanPb.setToCluster(reservedParam.remoteCluster);
        spanPb.setToDc(reservedParam.remoteDc);
        spanPb.setSendSize(reservedParam.sendSize);
        spanPb.setRecvSize(reservedParam.recvSize);
        spanPb.setToServiceType(reservedParam.serviceType);
        spanPb.setComponent(reservedParam.component);
        return spanPb;
    }
    _tranToCustomSpanProto() {
        const breaker = this.getBreaker();
        const { reservedParam } = this;
        const spanPb = new bytedtracePb.CustomSpanProto();
        if (this.reservedParam.isError) {
            this.tags.push([gConsts.IS_ERROR_KEY, true]);
        }
        spanPb.setTagsList(breaker.buildTags(this.tags) || []);
        spanPb.setRefType(this.spanRefType);
        spanPb.setReferencesList(this.buildSpanRefs());
        spanPb.setSpanId(this.spanId);
        spanPb.setParentId(this.parentSpanId);
        spanPb.setType(this.spanType);
        spanPb.setName(this.name);
        spanPb.setEventsList(this.genEventPbs());
        spanPb.setMetricsList(this.genMetricsProto());
        spanPb.setDuration(this.duration);
        spanPb.setStartTime(this.startTime);
        spanPb.setStatusCode(reservedParam.statusCode);
        return spanPb;
    }
    _resetReservedMetricsParam() {
        this.name = '-';
        this.spanType = '-';
        this.reservedParam.remoteService = '-';
        this.reservedParam.serviceType = '-';
        this.reservedParam.remoteMethod = '-';
        // set default cluster name to adapt to Alert System
        this.reservedParam.remoteCluster = consts.DEFAULT_CLUSTER;
        this.reservedParam.remoteDc = '-';
        this.reservedParam.component = '-';
        this.reservedParam.remoteAddrStr = '-';
    }
    _buildTraceLogSpan() {
        const log = {};
        if (this.reservedParam.isError) {
            log.level = 'error';
        }
        else {
            log.level = 'trace';
        }
        log.logId = this.getLogId();
        const tags = {
            psm: this.tracer.getServiceName(),
            spanId: this.spanId,
            parentSpanId: this.parentSpanId,
            costInUs: this.duration,
            remoteMethod: this.reservedParam.remoteMethod,
            remoteCluster: this.reservedParam.remoteCluster,
            remoteService: this.reservedParam.remoteService,
            remoteAddress: this.reservedParam.remoteAddrStr,
            statusCode: this.reservedParam.statusCode,
            startAt: this.startTime,
        };
        if (this.reservedParam.businessStatusCode !== 0) {
            tags.bizStatus = this.reservedParam.businessStatusCode;
        }
        if (this.reservedParam.sendSize !== 0) {
            tags.readSize = this.reservedParam.recvSize;
        }
        for (let i = 0; i < this.tags.length; i++) {
            const [key, val] = this.tags[i];
            tags[key] = val;
        }
        tags.sampled = this.isSampled();
        if (this.isServerSpan()) {
            log.location = consts.ACCESS_LOG_LOCATION;
            tags.localMethod = this.name;
            tags.type = consts.SERVER_SPAN_RPC_LOG_TYPE;
        }
        else {
            log.location = consts.CALL_LOG_LOCATION;
            tags.localMethod = this.getRootSpanMethod();
            tags.type = consts.CLIENT_SPAN_RPC_LOG_TYPE;
        }
        log.tags = tags;
        return log;
    }
    _buildTraceLogByCustomSpan() {
        const log = {
            level: 'trace',
            logId: this.getLogId(),
            location: consts.CUSTOM_SPAN_LOG_LOCATION,
        };
        const tags = {
            type: consts.CUSTOM_SPAN_LOG_TYPE,
            psm: this.tracer.getServiceName(),
            spanId: this.spanId,
            parentSpanId: this.parentSpanId,
            costInUs: this.duration,
            localMethod: this.spanType + '-' + this.name,
            statusCode: this.reservedParam.statusCode,
            startAt: this.startTime,
        };
        for (let i = 0; i < this.tags.length; i++) {
            const [key, val] = this.tags[i];
            tags[key] = val;
        }
        tags.sampled = this.isSampled();
        log.tags = tags;
        return log;
    }
    _samplingTrace(options = {
        samplingTraceType: consts.DEFAULT_SAMPLING_TRACE_OPTIONS.samplingTraceType,
        rootSpanContext: undefined,
    }) {
        const txn = this.transaction;
        switch (options.samplingTraceType) {
            case types.SamplingTraceType.localTraceType: {
                if (txn.isSampled() || txn.isLocalSampled()) {
                    return;
                }
                txn.setLocalSamplingTrace();
                break;
            }
            case types.SamplingTraceType.postTraceType: {
                if (txn.isPostTrace()) {
                    return;
                }
                if (!this.doPostTrace(options.rootSpanContext)) {
                    return;
                }
                if (options.rootSpanContext) {
                    return;
                }
                break;
            }
            case types.SamplingTraceType.debugTraceType: {
                if (txn.isDebugTrace()) {
                    return;
                }
                txn.setDebugTrace(types.SamplingTraceType.debugTraceType, true);
                break;
            }
            case types.SamplingTraceType.bothSideTraceType: {
                if (txn.isBothSideTrace()) {
                    return;
                }
                if (txn.isPostTrace()) {
                    txn.setDebugTrace(types.SamplingTraceType.bothSideTraceType, false);
                }
                else {
                    if (this.doPostTrace(options.rootSpanContext)) {
                        txn.setDebugTrace(types.SamplingTraceType.bothSideTraceType, false);
                    }
                    else {
                        return;
                    }
                }
                break;
            }
        }
        this.setTag(consts.SAMPLING_TRACE_ROOT_TYPE_TAG, options === null || options === void 0 ? void 0 : options.samplingTraceType);
    }
    /**
     * 增加执行次数
     * @param cnt
     * @param isSucc
     */
    incrExecCount(cnt, isSucc) {
        if (isSucc) {
            this.extInfo.succExecCnt += cnt;
        }
        else {
            this.extInfo.failExecCnt += cnt;
        }
    }
    /**
     * 记录执行一次
     * @param duration
     * @param isSucc
     */
    recordOneExec(duration, isSucc) {
        const extInfo = this.extInfo;
        if (isSucc) {
            extInfo.succExecCnt++;
            if (extInfo.maxSuccExecDuration < duration) {
                extInfo.maxSuccExecDuration = duration;
            }
        }
        else {
            extInfo.failExecCnt++;
            if (extInfo.maxFailExecDuration < duration) {
                extInfo.maxFailExecDuration = duration;
            }
        }
        const { metricsCounterName, metricsTimerName } = this._getMetricsName();
        if (this.enableMetrics()) {
            this.reportMetricsBase({
                metricsCounterName,
                metricsTimerName,
                isError: isSucc,
                latency: duration,
                methodName: this._getSpanMethodName(),
                cnt: 0,
                emitLatencyOnly: true,
                emitRateOnly: false,
                optionalMetrics: false,
            });
        }
    }
    enableMetrics() {
        const isCustom = this.isCustomSpan();
        if (!this.tracer.options.tracerConfig.isMetricsEnable(isCustom)) {
            return false;
        }
        if (typeof this.emitMetricsOpt === 'boolean') {
            return this.emitMetricsOpt;
        }
        return !isCustom;
    }
    buildSpanRefs() {
        var _a;
        const refereceModels = (_a = this.reference) === null || _a === void 0 ? void 0 : _a.map((ref) => {
            const spanRefModel = new bytedtracePb.SpanRef();
            spanRefModel.setTraceId(this.tranTraceIDLocal(ref.referencedContext.traceID));
            spanRefModel.setParentSpanId(ref.referencedContext.spanID);
            spanRefModel.setRefType(ref.referenceType);
            return spanRefModel;
        });
        return refereceModels || [];
    }
    tranTraceIDLocal(tracerId) {
        const traceIdModel = new bytedtracePb.TraceID();
        // https://github.com/protocolbuffers/protobuf/issues/3666
        const high = tracerId.high;
        const low = tracerId.low;
        traceIdModel.setHigh(high);
        traceIdModel.setLow(low);
        return traceIdModel;
    }
    genEventPbs() {
        const breaker = this.getBreaker();
        const eventPbs = this.events.map((event) => event.tranToPb(breaker));
        return eventPbs;
    }
    genMetricsProto() {
        var _a;
        const metricPbs = (_a = this.metricPbs) === null || _a === void 0 ? void 0 : _a.map((m) => m.tran());
        return metricPbs || [];
    }
    reportMetrics() {
        if (this.enableMetrics()) {
            this._reportMetrics();
        }
    }
    reportOptionalMetrics() {
        if (!this.tracer.options.enableSendRecvBytesMetrics &&
            this.extSpanModel !== gConsts.SpanModel.ConsumerSpanModel) {
            return;
        }
        if (this.tracer.options.enableSendRecvBytesMetrics) {
            if (this.reservedParam.sendSize > 0) {
                this.tracer.options.metricsClient.emitRateCounter(`span.${this.spanType}.send.bytes`, this.reservedParam.sendSize);
            }
            if (this.reservedParam.recvSize > 0) {
                this.tracer.options.metricsClient.emitRateCounter(`span.${this.spanType}.receive.bytes`, this.reservedParam.recvSize);
            }
        }
        if (this.extSpanModel === gConsts.SpanModel.ConsumerSpanModel) {
            if (this.extInfo && this.extInfo.mqLag !== 0) {
                this.tracer.options.metricsClient.emitTimer(`span.${this.spanType}.mq_log.us`, this.extInfo.mqLag * 1000);
            }
        }
    }
    _getSpanMethodName() {
        if (this.isClientSpan()) {
            return this.reservedParam.remoteMethod;
        }
        else {
            return this.name;
        }
    }
    _getMetricsName() {
        let metricsCounterName;
        let metricsTimerName;
        const methodName = this._getSpanMethodName();
        if (this.isServerSpan()) {
            metricsCounterName = consts.SERVICE_SPAN_METRICS_COUNTER_NAME;
            metricsTimerName = consts.SERVICE_SPAN_METRICS_TIMER_NAME;
        }
        else if (this.isClientSpan()) {
            metricsCounterName = consts.CLIENT_SPAN_METRICS_COUNTER_NAME;
            metricsTimerName = consts.CLIENT_SPAN_METRICS_TIMER_NAME;
        }
        else {
            metricsCounterName = 'span.' + this.spanType + '.rate';
            metricsTimerName = 'span.' + this.spanType + '.latency.us';
        }
        return {
            metricsCounterName,
            metricsTimerName,
            methodName,
        };
    }
    _reportMetrics() {
        const { metricsCounterName, metricsTimerName, methodName } = this._getMetricsName();
        switch (this.extSpanModel) {
            case gConsts.SpanModel.BatchSpanModel: {
                if (this.extInfo.succExecCnt > 0) {
                    this.reportMetricsBase({
                        metricsCounterName,
                        metricsTimerName,
                        methodName,
                        latency: 0,
                        cnt: this.extInfo.succExecCnt,
                        emitLatencyOnly: false,
                        emitRateOnly: true,
                        optionalMetrics: false,
                        isError: false,
                    });
                }
                if (this.extInfo.failExecCnt > 0) {
                    this.reportMetricsBase({
                        metricsCounterName,
                        metricsTimerName,
                        methodName,
                        latency: 0,
                        cnt: this.extInfo.failExecCnt,
                        emitLatencyOnly: false,
                        emitRateOnly: true,
                        optionalMetrics: false,
                        isError: true,
                    });
                }
                if (!this.extInfo.disableTotalMetrics) {
                    this.reportMetricsBase({
                        metricsCounterName,
                        metricsTimerName,
                        methodName: '_batch_' + methodName,
                        latency: this.duration,
                        cnt: 1,
                        emitLatencyOnly: false,
                        emitRateOnly: false,
                        optionalMetrics: true,
                        isError: this.reservedParam.isError,
                    });
                }
                break;
            }
            default: {
                this.reportMetricsBase({
                    metricsCounterName,
                    metricsTimerName,
                    methodName,
                    cnt: 1,
                    latency: this.duration,
                    emitLatencyOnly: false,
                    emitRateOnly: false,
                    optionalMetrics: true,
                    isError: this.reservedParam.isError,
                });
            }
        }
    }
    reportMetricsBase(options) {
        const tags = this.getSpanMetricTag(options);
        this.emitMetrics(tags, options);
    }
    emitMetrics(tags, options) {
        const { metricsClient } = this.tracer.options;
        const { metricsCounterName, metricsTimerName, cnt, emitLatencyOnly, emitRateOnly, optionalMetrics, latency, } = options;
        if (emitLatencyOnly) {
            metricsClient.emitTimer(metricsTimerName, latency, undefined, tags);
        }
        else if (emitRateOnly) {
            metricsClient.emitRateCounter(metricsCounterName, cnt, undefined, tags);
        }
        else {
            metricsClient.emitRateCounter(metricsCounterName, cnt, undefined, tags);
            metricsClient.emitTimer(metricsTimerName, latency, undefined, tags);
            if (optionalMetrics) {
                this.reportOptionalMetrics();
            }
        }
    }
    getSpanMetricTag(options) {
        const { methodName, isError } = options;
        const { tracerConfig, disableReportRemoteAddrWhenErr } = this.tracer.options;
        const spanMetricLevel = tracerConfig.getSpanMetricsLevel();
        const simplifyTags = tracerConfig.isSimplifySpanMetrics();
        let remoteAddrReport = '-';
        if (!simplifyTags &&
            (this.emitRemoteAddrInMetrics ||
                (!disableReportRemoteAddrWhenErr &&
                    this.reservedParam.isError &&
                    spanMetricLevel === 0))) {
            remoteAddrReport = this.reservedParam.remoteAddrStr;
        }
        const tags = {
            [gConsts.STATUS_CODE_KEY]: this.reservedParam.statusCode,
            [gConsts.IS_ERROR_KEY]: isError ? '1' : '0',
        };
        if (this.spanType === gConsts.SERVER_SPAN_TYPE) {
            tags[gConsts.SERVICE_TYPE_KEY] = this.reservedParam.serviceType;
            tags[gConsts.FROM_SERVICE_KEY] = this.reservedParam.remoteService;
            tags[gConsts.FROM_CLUSTER_KEY] = this.reservedParam.remoteCluster;
            tags[gConsts.FROM_DC_KEY] = this.reservedParam.remoteDc;
            tags[gConsts.FROM_REMOTE_ADDR] = remoteAddrReport;
            tags[gConsts.BUSINESS_STATUS_CODE_KEY] =
                this.reservedParam.businessStatusCode;
            tags[gConsts.METHOD_KEY] = methodName;
            if (this.reservedParam.serviceType === 'http') {
                this.tags.forEach((tag) => {
                    if (tag[0] === gConsts.HTTP_ROUTE) {
                        tags[gConsts.HTTP_ROUTE] = tag[1];
                    }
                    if (tag[0] === gConsts.HTTP_METHOD) {
                        tags[gConsts.HTTP_METHOD] = tag[1];
                    }
                });
            }
            const registerTags = gUtils.fillExtraMetricsTags(this.tags, this.tracer.options.registerServerSpanMetricTags, simplifyTags);
            return Object.assign(Object.assign({}, tags), registerTags);
        }
        else if (this.spanType === gConsts.CLIENT_SPAN_TYPE) {
            tags[gConsts.TO_SERVICE_KEY] = this.reservedParam.remoteService;
            tags[gConsts.TO_METHOD_KEY] = this.reservedParam.remoteMethod;
            tags[gConsts.TO_SERVICE_TYPE_KEY] = this.reservedParam.serviceType;
            tags[gConsts.TO_CLUSTER_KEY] = this.reservedParam.remoteCluster;
            tags[gConsts.TO_DC_KEY] = this.reservedParam.remoteDc;
            tags[gConsts.TO_REMOTE_ADDR] = remoteAddrReport;
            tags[gConsts.BUSINESS_STATUS_CODE_KEY] =
                this.reservedParam.businessStatusCode;
            if (simplifyTags) {
                tags[gConsts.METHOD_KEY] = '-';
            }
            else {
                tags[gConsts.METHOD_KEY] = this.getRootSpanMethod();
            }
            // TODO: hotfix support http/rpc client emit metrics `request.env`
            // needs to be refactored to handle all defined tags according to the bytedtrace specification
            this.tracer.appendSpanMetricTags(gTypes.SpanType.client, 'request.env');
            const registerTags = gUtils.fillExtraMetricsTags(this.tags, this.tracer.options.registerClientSpanMetricTags, simplifyTags);
            return Object.assign(Object.assign({}, tags), registerTags);
        }
        // customSpan
        tags[gConsts.NAME_KEY] = this.name;
        const registerTags = gUtils.fillExtraMetricsTags(this.tags, this.tracer.options.registerCustomSpanMetricTags, false);
        return Object.assign(Object.assign({}, tags), registerTags);
    }
    isCustomSpan() {
        return !(this.isServerSpan() || this.isClientSpan());
    }
    emitSpanLog() {
        if (this.enableEmitSpanLog()) {
            this.tracer.emitSpanLog(this);
        }
    }
    enableEmitSpanLog() {
        if (typeof this.emitSpanLogOpt === 'boolean') {
            return this.emitSpanLogOpt;
        }
        return gConsts.DEFAULT_ENABLE_SPAN_LOG;
    }
    changeSpanId(id) {
        this.setTag(gConsts.ORIGIN_SPAN_ID_KEY, this.spanId);
        this.spanId = id;
    }
    applyFinish() {
        if (this.isFinished === consts.SPAN_UN_FINISHED) {
            this.isFinished = consts.SPAN_FINISHED;
            return true;
        }
        /* istanbul ignore next */
        return false;
    }
    applyPostTrace() {
        if (this.postTrace === 0) {
            this.postTrace = 1;
            return true;
        }
        return false;
    }
    genPostTraceLimitName() {
        return this.spanType + gConsts.LINKER + this.name;
    }
    doPostTrace(ctx) {
        const txn = this.transaction;
        const { breaker, statusCounter } = txn.tracer.options;
        if (txn.isSampled()) {
            return false;
        }
        const limitPost = this.genPostTraceLimitName();
        if (!breaker.postTraceLimiterPack.pass(limitPost)) {
            statusCounter.recordWarningWithRateLimit(false, gStatus.ErrPostTraceOverflow, `post trace over limit, operation:${limitPost}`);
            return false;
        }
        if (!this.applyPostTrace()) {
            return false;
        }
        if (ctx) {
            if (!(ctx instanceof spanContext_1.SpanContext)) {
                statusCounter.recordWarningWithRateLimit(false, gStatus.ErrPostTraceCtxInvalid, 'post trace err,context is not spanContext.');
                return false;
            }
            else {
                const success = txn.setPostTrace(ctx);
                if (success) {
                    let nId;
                    if (ctx.parentID === '0') {
                        nId = (BigInt(ctx.spanID) - BigInt(1)).toString(10);
                    }
                    else {
                        nId = ctx.parentID;
                    }
                    if (this === txn.rootSpan) {
                        txn.rootSpanId = nId;
                    }
                    this.changeSpanId(nId);
                    return true;
                }
                return false;
            }
        }
        else {
            return txn.setPostTrace();
        }
    }
    saveMetrics(mType, name, value, tagKvs) {
        this.addMetrics({ value, tagKvs, name, mType });
    }
    addMetrics(metricPb) {
        if (this.metricPbs.length >=
            this.getBreaker().getMaxCustomMetricCountInSpan()) {
            this.tracer.options.statusCounter.recordWarningWithRateLimit(false, gStatus.ErrMetricNumInSpanOverflow, 'custom metrics number in span exceed cap!');
            return false;
        }
        const _metricPb = Object.assign({}, metricPb);
        _metricPb.tracer = this.tracer;
        this.metricPbs.push(new metricPb_1.MetricPb(_metricPb));
        return true;
    }
    shouldReported() {
        return this.transaction.shouldReported();
    }
}
exports.Span = Span;
