// @byted-service/bytedtrace/lib/span/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.samplingMethodMap = exports.SamplingTraceType = void 0;
// https://bytedance.feishu.cn/docs/doccnDVjsjb6levh5fId8dSICUg
var SamplingTraceType;
(function (SamplingTraceType) {
    /**
     * SamplingTrace only happen in local txn
     * 0b0010
     */
    SamplingTraceType[SamplingTraceType["localTraceType"] = 1] = "localTraceType";
    /**
     * SamplingTrace happen in local txn and transfer to backward
     * 0b0101
     */
    SamplingTraceType[SamplingTraceType["postTraceType"] = 2] = "postTraceType";
    /**
     * SamplingTrace happen in local txn and transfer to forward
     * 0b0011
     */
    SamplingTraceType[SamplingTraceType["debugTraceType"] = 3] = "debugTraceType";
    /**
     * SamplingTrace happen in local txn and transfer to backward and forward
     * 0b0111
     */
    SamplingTraceType[SamplingTraceType["bothSideTraceType"] = 4] = "bothSideTraceType";
})(SamplingTraceType = exports.SamplingTraceType || (exports.SamplingTraceType = {}));
exports.samplingMethodMap = {};
