// @universe-design/react/lib/tree/common.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UD_TREE_NODE_ID = void 0;
exports.treeChildren2TreeNodes = treeChildren2TreeNodes;
exports.useTreeCssScope = useTreeCssScope;
var _hooks = require("../config-provider/hooks");
var _util = require("../util");
function useTreeCssScope() {
  return (0, _hooks.useCssScope)('tree');
}
const UD_TREE_NODE_ID = exports.UD_TREE_NODE_ID = 'UD_TREE_NODE';
function isTreeNode(node) {
  var _node$type;
  return node === null || node === void 0 ? void 0 : (_node$type = node.type) === null || _node$type === void 0 ? void 0 : _node$type[UD_TREE_NODE_ID];
}
function treeChildren2TreeNodes(children) {
  return (0, _util.toFlatArray)(children).map((child, index) => {
    if (isTreeNode(child)) {
      const {
        key,
        props: {
          children: _children,
          ...restProps
        }
      } = child;
      return {
        ...restProps,
        children: treeChildren2TreeNodes(_children),
        key: key || restProps.value || `tree-node-key-${index}`
      };
    }
    return null;
  }).filter(_util.isNonNullable);
}
