// @universe-design/react/lib/tree/TreeNodeLabel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TreeNodeLabel = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _shared = require("../shared");
var _TreeContext = require("./hooks/TreeContext");
var _TreeNodeContext = require("./hooks/TreeNodeContext");
var _util = require("../util");
var _textOverflow = require("../shared/components/text-overflow");
const TreeNodeLabel = exports.TreeNodeLabel = /*#__PURE__*/_react.default.forwardRef((_ref, ref) => {
  let {
    className,
    children,
    ...restProps
  } = _ref;
  const {
    rootCssScope,
    labelRenderer
  } = (0, _TreeContext.useTreeContext)();
  const {
    doublyTreeNode: node
  } = (0, _TreeNodeContext.useTreeNodeContext)();
  const {
    disabled,
    label,
    value,
    title
  } = node.__node__;
  const markRenderer = _react.default.useCallback(text => /*#__PURE__*/_react.default.createElement("span", {
    className: (0, _clsx.default)(!disabled && rootCssScope('__node-highlight'))
  }, text), [disabled, rootCssScope]);
  const labelNode = (() => {
    if ((0, _util.isNonNullable)(children)) {
      return children;
    }
    if ((0, _util.isNonNullable)(label)) {
      return label;
    }
    if ((0, _util.isNonNullable)(title)) {
      return title;
    }
    if (labelRenderer) {
      return labelRenderer(node);
    }
    return value;
  })();
  const $labelNode = /*#__PURE__*/_react.default.createElement(_shared.HighlightConsumer, {
    markRenderer: markRenderer
  }, labelNode);
  return /*#__PURE__*/_react.default.createElement(_textOverflow.TextOverflow, (0, _extends2.default)({
    ref: ref,
    as: "div",
    className: (0, _clsx.default)(rootCssScope('__node__label'), className),
    lineClamp: 1
  }, restProps), $labelNode);
});
