// @universe-design/react/lib/tree/TreeNodeCheckbox.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TreeNodeCheckbox = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _TreeContext = require("./hooks/TreeContext");
var _interface = require("./interface");
var _Checkbox = require("../checkbox/Checkbox");
var _TreeNodeContext = require("./hooks/TreeNodeContext");
var _util = require("./util");
var _composeRefs = _interopRequireDefault(require("../util/composeRefs"));
var _util2 = require("../util");
const TreeNodeCheckbox = exports.TreeNodeCheckbox = /*#__PURE__*/_react.default.forwardRef((_ref, ref) => {
  let {
    // to be merged
    className,
    onChange,
    // to be omitted
    disabled,
    ...restProps
  } = _ref;
  const {
    checkable,
    toggleChecked,
    rootCssScope
  } = (0, _TreeContext.useTreeContext)();
  const {
    doublyTreeNode: node,
    checkboxRef,
    isChecked
  } = (0, _TreeNodeContext.useTreeNodeContext)();

  /**
   * 合并 contextRef 和 userRef
   */
  const mergedRef = _react.default.useMemo(() => (0, _composeRefs.default)(checkboxRef, ref), [checkboxRef, ref]);
  const {
    value
  } = node.__node__;
  const checkedDisabled = (0, _util.isCheckedDisabled)(node.__node__);
  let checkboxOverlapProps = {
    disabled: disabled !== null && disabled !== void 0 ? disabled : checkedDisabled,
    className: (0, _clsx.default)(rootCssScope('__node__checkbox'), className),
    onChange: _react.default.useCallback(e => {
      (0, _util2.isFunction)(onChange) && onChange(e);
      toggleChecked(value);
    }, [onChange, toggleChecked, value])
  };
  if (isChecked === _interface.CheckedType.Checked) {
    checkboxOverlapProps = {
      checked: true,
      indeterminate: false,
      ...checkboxOverlapProps
    };
  } else if (isChecked === _interface.CheckedType.Unchecked) {
    checkboxOverlapProps = {
      checked: false,
      indeterminate: false,
      ...checkboxOverlapProps
    };
  } else if (isChecked === _interface.CheckedType.Indeterminate) {
    checkboxOverlapProps = {
      checked: false,
      indeterminate: true,
      ...checkboxOverlapProps
    };
  }
  if (checkable) {
    return /*#__PURE__*/_react.default.createElement(_Checkbox.Checkbox, (0, _extends2.default)({
      ref: mergedRef
    }, checkboxOverlapProps, restProps));
  }
  return null;
});
