// @universe-design/react/lib/tree/TreeContainer.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TreeContainer = void 0;
var _rcVirtualList = _interopRequireDefault(require("@universe-design/rc-virtual-list"));
var React = _interopRequireWildcard(require("react"));
var _TreeContext = require("./hooks/TreeContext");
var _InternalTreeNode = require("./InternalTreeNode");
var _shared = require("../shared");
var _list = require("../shared/components/list");
const TreeContainer = () => {
  const {
    finalTreeNodeList,
    rootCssScope,
    height,
    virtual,
    listRef,
    searchValue,
    notFoundContent
  } = (0, _TreeContext.useTreeContext)();
  const listNode = /*#__PURE__*/React.createElement("div", {
    className: rootCssScope('__list')
  }, /*#__PURE__*/React.createElement(_rcVirtualList.default, {
    itemKey: node => node.__node__.value,
    ref: listRef,
    data: finalTreeNodeList,
    height: height,
    itemHeight: 32,
    fullHeight: false
    // onScroll={onPopupScroll}
    ,
    virtual: virtual
    // onMouseEnter={onMouseEnter}
  }, node =>
  /*#__PURE__*/
  // 这里加一个空 div 是为了把内部带 margin 的元素包住，让 virtual list 能正常计算，tree，select，cascader
  React.createElement("div", null, /*#__PURE__*/React.createElement(_InternalTreeNode.InternalTreeNode, {
    key: node.__node__.value,
    doublyTreeNode: node
  }))));
  const notFoundNode = (() => {
    if (notFoundContent === null) {
      return null;
    }
    return /*#__PURE__*/React.createElement(_list.ListItemLike, {
      className: rootCssScope('__empty')
    }, notFoundContent);
  })();
  return /*#__PURE__*/React.createElement(_shared.HighlightProvider, {
    value: React.useMemo(() => ({
      searchWords: searchValue ? [searchValue] : undefined,
      disabled: false
    }), [searchValue])
  }, /*#__PURE__*/React.createElement("div", {
    className: rootCssScope()
  }, finalTreeNodeList.length ? listNode : notFoundNode));
};
exports.TreeContainer = TreeContainer;
