// @universe-design/react/lib/tree/TreeNodeExpandButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TreeNodeExpandButton = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _useTreeNodeLoadData = require("./hooks/useTreeNodeLoadData");
var _TreeNodeContext = require("./hooks/TreeNodeContext");
var _TreeContext = require("./hooks/TreeContext");
var _util = require("../util");
var _composeRefs = _interopRequireDefault(require("../util/composeRefs"));
var _expandButton = require("../shared/components/expand-button");
const TreeNodeExpandButton = exports.TreeNodeExpandButton = /*#__PURE__*/_react.default.forwardRef((_ref, ref) => {
  let {
    className,
    onClick,
    ...restProps
  } = _ref;
  const {
    rootCssScope,
    switcherIcon,
    finalTreeNodeList,
    loadData,
    loadedKeys,
    onLoad,
    setLoadedKeys,
    toggleExpanded
  } = (0, _TreeContext.useTreeContext)();
  const {
    doublyTreeNode: node,
    expandButtonRef,
    isExpanded
  } = (0, _TreeNodeContext.useTreeNodeContext)();
  const mergedRef = _react.default.useMemo(() => (0, _composeRefs.default)(ref, expandButtonRef), [expandButtonRef, ref]);
  const {
    onInternalExpandedToggle,
    loading,
    isLeaf
  } = (0, _useTreeNodeLoadData.useTreeNodeLoadData)({
    doublyTreeNode: node,
    loadData,
    loadedKeys,
    onLoad,
    setLoadedKeys,
    toggleExpanded
  });

  // 如果是第一层，并且 root 下没有 non-leaf 的节点，这时其实所有节点都是没有 expand icon 的所以可以不显示 empty-expand-icon
  const ifRootHasNoNonLeafChildren = finalTreeNodeList.filter(o => o.depth === 1).every(o => (0, _useTreeNodeLoadData.isLeafWithLoadData)(o.__node__, {
    loadData,
    loadedKeys
  }));
  if (node.depth === 1 && ifRootHasNoNonLeafChildren) {
    return null;
  }
  return /*#__PURE__*/_react.default.createElement(_expandButton.ExpandButton, (0, _extends2.default)({
    ref: mergedRef,
    className: (0, _clsx.default)(rootCssScope('__node__expandIcon'), className),
    loading: loading,
    isExpanded: isExpanded,
    onClick: e => {
      (0, _util.isFunction)(onClick) && onClick(e);
      onInternalExpandedToggle();
    },
    icon: switcherIcon,
    iconType: "ExpandDownFilled",
    asPlaceholder: isLeaf
  }, restProps));
});
