// @universe-design/react/es/badge/CanvasTextBadge.js
import React from 'react';
import { function_danger_fill_default, static_white } from '@universe-design/styles/esm/all.light.atom';
import { useTokens } from '@universe-design/styles';
import { useMedia } from '../hooks';
import { isClient } from '@universe-design/react-utilities';
import omit from 'lodash/omit';
import { defaults, getBadgeFont, isPresetColor } from './utils';
function getScale() {
  return isClient() ? window.devicePixelRatio : 1;
}
export function createCanvasTextBadgeSidercar() {
  let {
    size = defaults.textBadgeFontSize,
    fontSize = 10,
    enableMediaQueryResolutionMonitor = false,
    textVerticalAlign = 'baseline',
    deviceScaleFactor = 1
  } = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return _ref => {
    var _tokens$staticWhite, _tokens$functionDang;
    let {
      color: backgroundColorProp,
      content,
      className,
      style
    } = _ref;
    const tokens = useTokens();
    const color = (_tokens$staticWhite = tokens['static-white']) !== null && _tokens$staticWhite !== void 0 ? _tokens$staticWhite : static_white;
    const backgroundColor = isPresetColor(backgroundColorProp) ? (_tokens$functionDang = tokens['function-danger-fill-default']) !== null && _tokens$functionDang !== void 0 ? _tokens$functionDang : function_danger_fill_default : backgroundColorProp;
    const canvasRef = React.useRef(null);
    const scale = getScale() * deviceScaleFactor;
    useMedia(`(resolution: ${scale}dppx)`, {
      skipped: !enableMediaQueryResolutionMonitor
    });
    React.useLayoutEffect(function () {
      function floor(n) {
        return Math.floor(n);
      }
      const canvas = canvasRef.current;
      if (!canvas) {
        return;
      }
      const ch = size * scale;
      canvas.height = ch;
      const d = ch;
      const r = d / 2;
      const ctx = canvas.getContext('2d');
      if (!ctx) {
        return;
      }
      /**
       * scale 画布
       */
      // 抗锯齿 https://stackoverflow.com/questions/195262/can-i-turn-off-antialiasing-on-an-html-canvas-element
      // ctx.translate(0.5, 0.5);
      ctx.scale(scale, scale);
      const font = getBadgeFont(fontSize * scale);
      const canvasPadding = 0;

      /**
       * 算各种尺寸
       */
      ctx.font = font;
      // 行高
      const metrics = ctx.measureText(content);
      // 这里不能用 right - left ，因为阿拉伯语是反过来的
      const tw = metrics.width;
      const th = metrics.actualBoundingBoxAscent + metrics.actualBoundingBoxDescent;
      const paddingX = r * 4 / 7;
      const isCircle = tw < r * 2 - paddingX * 2 || [...content].length === 1;
      const cw = (() => {
        if (isCircle) {
          return d;
        }
        return tw + paddingX * 2;
      })();

      // set width
      canvas.style.height = `${size}px`;
      canvas.style.width = `${cw / scale}px`;
      canvas.width = cw;

      /**
       * 画左半圆
       */
      ctx.beginPath();
      ctx.arc(floor(r), floor(r), floor(r - canvasPadding), 0, Math.PI * 2);
      ctx.fillStyle = backgroundColor;
      ctx.fill();

      /**
       * 画右半圆
       */
      ctx.beginPath();
      ctx.arc(floor(cw - r), floor(r), floor(r - canvasPadding), 0, Math.PI * 2);
      ctx.fillStyle = backgroundColor;
      ctx.fill();

      /**
       * 画中间矩形
       */
      if (!isCircle) {
        ctx.beginPath();
        ctx.fillStyle = backgroundColor;
        ctx.fillRect(floor(r), floor(canvasPadding), floor(cw - r * 2), floor(ch - canvasPadding * 2));
      }

      /**
       * 写字
       */
      ctx.textAlign = 'center';
      let ty;
      if (textVerticalAlign === 'baseline') {
        // 第一种是 font middle，相当于 align-items: baseline
        ctx.textBaseline = 'middle';
        ty = ch / 2;
      } else {
        // 第二种是 actual middle，相当于 align-items: center
        ctx.textBaseline = 'alphabetic';
        ty = (ch + th) / 2 - metrics.actualBoundingBoxDescent;
      }
      // console.log({ content, cw, ch, tw, th, r, d, font, metrics });
      ctx.fillStyle = color;
      ctx.font = font;
      ctx.fillText(content, floor(cw / 2), floor(ty));
      return () => {
        ctx.clearRect(0, 0, cw, ch);
      };
    }, [backgroundColor, color, content, scale]);
    return /*#__PURE__*/React.createElement("canvas", {
      style: omit(style, ['width', 'height']),
      className: className,
      ref: canvasRef
    });
  };
}
export const CanvasTextBadgeSidecar = createCanvasTextBadgeSidercar();
