// @universe-design/react/es/badge/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.u1xvp9sr{transform:none;}
import uko6l5r from "./../.stylex/uko6l5r";
//.u10pi13n{position:relative;}
import u5jxv71 from "./../.stylex/u5jxv71";
//.u1euv43f{position:absolute;}
import u1u1p7mj from "./../.stylex/u1u1p7mj";
//.u1wirg6b{z-index:auto;}
import u1r95smd from "./../.stylex/u1r95smd";
//.uqzzkal{font-family:"PingFang SC","Microsoft YaHei","Myriad Pro","Hiragino Sans GB",sans-serif;}
import ulb4spa from "./../.stylex/ulb4spa";
//.unmn6fi{font-size:10px;}
import u16uuw9n from "./../.stylex/u16uuw9n";
/* eslint-disable max-lines-per-function */
/* eslint-disable @typescript-eslint/indent */
import MoreOutlined from '@universe-design/icons-react/MoreOutlined';
import clsx from 'clsx';
import React from 'react';
import stylex from '@universe-design/stylex';
import { cssBlockBase, cssFontMedium } from '../stylex';
import { isNonNullable } from '../util';
import composeRefs from '../util/composeRefs';
import { useProxyRef, useControlledState } from '@universe-design/react-utilities';
import { getBadgeRadius, getBorderRadius, isPresetColor, normalizeCssSize, useBadge } from './utils';
import './style';
const MAX_BADGE_DISPLAY_NUMBER = 999;
const BADGE_FONT_SIZE = 10;
const cssBadgeBadge = stylex.compose(cssBlockBase, cssFontMedium, {
  "props": {
    "qhf8xq": u1u1p7mj,
    "Bj3rh1h": u1r95smd,
    "Bahqtrf": ulb4spa,
    "Be2twd7": u16uuw9n
  }
});
export const BaseBadge = /*#__PURE__*/React.forwardRef((props, ref) => {
  const {
    children,
    status,
    offset,
    style: styleProp,
    className,
    title
  } = props;
  let {
    count
  } = props;
  const {
    badgeCssScope,
    shape,
    position,
    isHidden,
    showAsDot,
    hasStatus
  } = useBadge(props);

  // Title
  const titleNode = title !== null && title !== void 0 ? title : typeof count === 'string' || typeof count === 'number' ? String(count) : undefined;
  if (isHidden) {
    return /*#__PURE__*/React.createElement(React.Fragment, null, children);
  }
  const hasChild = Boolean(children);
  let style = {
    ...styleProp
  };
  let colorClassName = '';
  const badgeBadgeCls = clsx(badgeCssScope('__badge'), !children && badgeCssScope('-has-no-child'), stylex(cssBadgeBadge, !children && {
    "props": {
      "Bz10aip": uko6l5r,
      "qhf8xq": u5jxv71
    }
  }));
  if (hasChild) {
    /**
     * 📢 定位要用 translate3d，在某些场景下可能出现 badge 内容左右偏移，加这个可以让 badge 独立一层
     */
    if (position === 'topRight') {
      style = {
        top: 0,
        right: 0,
        transform: offset ? `translate3d(calc(1 * (${normalizeCssSize(offset[0])})), calc(-1 * (${normalizeCssSize(offset[1])})), 0)` : undefined,
        ...style
      };
    } else if (position === 'topLeft') {
      style = {
        top: 0,
        left: 0,
        transform: offset ? `translate3d(calc(-1 * (${normalizeCssSize(offset[0])})), calc(-1 * (${normalizeCssSize(offset[1])})), 0)` : undefined,
        ...style
      };
    } else if (position === 'bottomRight') {
      style = {
        bottom: 0,
        right: 0,
        transform: offset ? `translate3d(calc(1 * (${normalizeCssSize(offset[0])})), calc(1 * (${normalizeCssSize(offset[1])})), 0)` : undefined,
        ...style
      };
    } else {
      style = {
        bottom: 0,
        left: 0,
        transform: offset ? `translate3d(calc(-1 * (${normalizeCssSize(offset[0])})), calc(1 * (${normalizeCssSize(offset[1])})), 0)` : undefined,
        ...style
      };
    }
  }
  if (showAsDot) {
    // Dot badge, Status badge
    const {
      colorStyle = 'filled',
      color,
      text,
      size
    } = props;
    const radius = getBadgeRadius({
      size,
      dot: true
    });
    style = {
      width: radius * 2,
      height: radius * 2,
      flexShrink: 0,
      ...style
    };
    if (color) {
      // color 优先级高于 status
      if (isPresetColor(color)) {
        colorClassName = clsx(badgeCssScope(`-${color}-${colorStyle}`), badgeCssScope(`-${colorStyle}`));
      } else if (colorStyle === 'filled') {
        style = {
          backgroundColor: color,
          ...style
        };
      } else {
        style = {
          backgroundColor: 'transparent',
          border: `1px solid ${color}`,
          ...style
        };
      }
    } else if (status) {
      colorClassName = clsx(badgeCssScope(`-${status}-${colorStyle}`), badgeCssScope(`-${colorStyle}`));
    } else {
      colorClassName = clsx(badgeCssScope(`-red-${colorStyle}`), badgeCssScope(`-${colorStyle}`));
    }
    if (!hasStatus) {
      return /*#__PURE__*/React.createElement("span", {
        className: clsx(badgeCssScope(), badgeCssScope('-dot'), className)
      }, children, /*#__PURE__*/React.createElement("sup", {
        ref: ref,
        title: titleNode,
        className: clsx(badgeBadgeCls, badgeCssScope(`-${position}`), colorClassName),
        style: style
      }));
    } else {
      return /*#__PURE__*/React.createElement("span", {
        className: clsx(badgeCssScope(), badgeCssScope('-dot-inline'), className)
      }, /*#__PURE__*/React.createElement("span", {
        ref: ref,
        className: clsx(badgeBadgeCls, colorClassName),
        style: style
      }), isNonNullable(text) && /*#__PURE__*/React.createElement("span", {
        className: clsx(badgeCssScope('__text'))
      }, text));
    }
  } else {
    // Text badge
    let type = 'text';
    const {
      overflowCount = MAX_BADGE_DISPLAY_NUMBER,
      color = 'red',
      hack10px = false,
      textBadgeSidecar: TextBadgeSidecar
    } = props;

    // 处理 Sidecar 逻辑
    const supClassName = clsx(badgeBadgeCls, badgeCssScope(`-${position}`));
    const isCountOverflow = typeof count === 'number' && count > overflowCount;
    if (TextBadgeSidecar && typeof count !== 'object' && !isCountOverflow) {
      return /*#__PURE__*/React.createElement("span", {
        className: clsx(badgeCssScope(), className)
      }, children, /*#__PURE__*/React.createElement(TextBadgeSidecar, {
        color: color,
        className: supClassName,
        style: style,
        content: String(count)
      }));
    }
    if (isPresetColor(color)) {
      colorClassName = badgeCssScope(`-${color}-filled`);
    } else {
      style = {
        backgroundColor: color,
        ...style
      };
    }
    if (shape === 'circle' || String(count).length === 1) {
      type = 'icon';
    }
    if (isCountOverflow) {
      count = /*#__PURE__*/React.createElement(MoreOutlined, {
        className: badgeCssScope('__badge__more')
      });
    }
    return /*#__PURE__*/React.createElement("span", {
      className: clsx(badgeCssScope(), badgeCssScope(`-${type}`), className)
    }, children, /*#__PURE__*/React.createElement("sup", {
      ref: ref,
      title: titleNode,
      className: clsx(supClassName, colorClassName),
      style: style
    }, /*#__PURE__*/React.createElement("div", {
      className: clsx(badgeCssScope('__badge__content'), hack10px && badgeCssScope('__badge__content-hack-10px'))
    }, count)));
  }
});
export const Badge = /*#__PURE__*/React.forwardRef((props, ref) => {
  const {
    children,
    offset: offsetProp,
    position,
    contentType,
    childCornerRadius: childCornerRadiusProp,
    ...restProps
  } = props;
  const {
    showAsDot
  } = useBadge(props);
  const shouldGetChildrenBorderRadius = !isNonNullable(childCornerRadiusProp) && !isNonNullable(offsetProp);
  const [childCornerRadius, setChildCornerRadius] = useControlledState(0, childCornerRadiusProp);
  const childRef = useProxyRef(null, node => {
    if (node && node instanceof HTMLElement && shouldGetChildrenBorderRadius) {
      setChildCornerRadius(getBorderRadius(node, position));
    }
  });
  const mergedChildRef = React.useMemo(() => composeRefs(childRef, children === null || children === void 0 ? void 0 : children.ref), [childRef, children]);
  const badgeRadius = getBadgeRadius(props);
  const badgeCenterPoint = {
    x: badgeRadius,
    y: badgeRadius
  };
  const tangencyPoint = {
    x: (1 - Math.SQRT1_2) * childCornerRadius,
    y: (1 - Math.SQRT1_2) * childCornerRadius
  };
  // Offset
  const offset = React.useMemo(() => {
    let _offset = [-tangencyPoint.x + badgeCenterPoint.x, -tangencyPoint.y + badgeCenterPoint.y];
    if (contentType === 'text' && showAsDot) {
      _offset = ['2px + 100%', 0];
    } else if (contentType === 'text') {
      _offset = ['2px + 100%', '50%'];
    } else if (isNonNullable(offsetProp)) {
      _offset = offsetProp;
    }
    return _offset;
  }, [tangencyPoint.x, tangencyPoint.y, badgeCenterPoint.x, badgeCenterPoint.y, contentType, showAsDot, offsetProp]);

  // 如果是有 offsetProp 或者 children 没法 cloneELement 或者不需要获取子元素圆角的时候直接返回 base badge
  if (! /*#__PURE__*/React.isValidElement(children) || !shouldGetChildrenBorderRadius) {
    return /*#__PURE__*/React.createElement(BaseBadge, _extends({}, restProps, {
      ref: ref,
      offset: offset,
      position: position
    }), children);
  }
  return /*#__PURE__*/React.createElement(BaseBadge, _extends({}, restProps, {
    ref: ref,
    offset: offset,
    position: position
  }), /*#__PURE__*/React.cloneElement(children, {
    ref: mergedChildRef
  }));
});
export * from './SVGTextBadge';
export * from './CanvasTextBadge';
export * from './types';
