// @universe-design/react/es/badge/utils.js
import { isClient } from '@universe-design/react-utilities';
import { useCssScope } from '../config-provider/hooks';
import { isNonNullable, PresetColorEnum } from '../util';
export const defaults = {
  textBadgeFontSize: 15
};
const EMPTY_TEXT_METRICS = {
  actualBoundingBoxAscent: 0,
  actualBoundingBoxDescent: 0,
  actualBoundingBoxLeft: 0,
  actualBoundingBoxRight: 0,
  width: 0,
  fontBoundingBoxAscent: 0,
  fontBoundingBoxDescent: 0,
  alphabeticBaseline: 0,
  emHeightAscent: 0,
  emHeightDescent: 0,
  hangingBaseline: 0,
  ideographicBaseline: 0
};
let ctxForMeasure = null;
export const measureText = (content, font) => {
  if (!content.length || !isClient()) {
    return EMPTY_TEXT_METRICS;
  }
  if (!ctxForMeasure) {
    ctxForMeasure = document.createElement('canvas').getContext('2d');
  }
  if (!ctxForMeasure) {
    return EMPTY_TEXT_METRICS;
  }
  ctxForMeasure.font = font;
  const dim = ctxForMeasure.measureText(content);
  return dim;
};
const PresetColorRegex = new RegExp(`^(${PresetColorEnum.join('|')})$`);
export const isPresetColor = color => PresetColorRegex.test(color);
export const getBadgeFont = fontSize => `500 ${fontSize}px "PingFang SC", "Microsoft YaHei", "Myriad Pro", "Hiragino Sans GB", sans-serif`;

/**
 * 根据 prop 计算 badge 的半径
 */
export const getBadgeRadius = _ref => {
  let {
    size = 'medium',
    dot = false
  } = _ref;
  if (!dot) {
    return 7;
  }
  let radius = 4;
  if (typeof size === 'number') {
    radius = size / 2;
  } else if (size === 'large') {
    radius = 5;
  } else if (size === 'small') {
    radius = 3;
  }
  return radius;
};

/**
 * 获取方形或圆形头像的 border radius px 值
 * 不适用于通用场景
 */
export function getBorderRadius(el, position) {
  const css = getComputedStyle(el);

  // webit-border-radius 为兼容 firefox
  let borderRadiusString = css.borderTopRightRadius || css.webkitBorderTopRightRadius;
  if (position === 'bottomLeft') {
    borderRadiusString = css.borderBottomLeftRadius || css.webkitBorderBottomLeftRadius;
  } else if (position === 'bottomRight') {
    borderRadiusString = css.borderBottomRightRadius || css.webkitBorderBottomRightRadius;
  } else if (position === 'topLeft') {
    borderRadiusString = css.borderTopLeftRadius || css.webkitBorderTopLeftRadius;
  }
  const borderRadius = parseFloat(borderRadiusString);
  if (borderRadiusString.match(/^\d+(?:\.\d+)?%$/)) {
    return Math.min(el.offsetHeight * borderRadius / 100, el.offsetHeight / 2);
  }
  if (Object.is(borderRadius, NaN)) {
    return 0;
  }
  return Math.min(borderRadius, el.offsetHeight / 2);
}
export function useBadge(props) {
  const {
    status,
    position = 'topRight',
    showZero = false,
    dot = false,
    shape = 'round',
    count
  } = props;
  const badgeCssScope = useCssScope('badge');
  // ======== 如果不 visible 直接返回 children  ========
  const hasStatus = isNonNullable(status) || props.text;
  const isZero = count === '0' || count === 0;
  /**
   * 显示为点状徽标还是文字徽标
   */
  const showAsDot = dot && !isZero || hasStatus;
  const isEmpty = !isNonNullable(count) || count === '';
  /**
   * 是否隐藏徽标
   */

  return {
    position,
    shape,
    badgeCssScope,
    hasStatus,
    showAsDot,
    isHidden: (isEmpty || isZero && !showZero) && !showAsDot
  };
}
export function normalizeCssSize(s) {
  return typeof s === 'number' ? `${s}px` : s;
}
