// @universe-design/react-provider/lib-commonjs/date-adapters/dayjsAdapter.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.dayjsAdapter = void 0;
var _dayjs = _interopRequireDefault(require("dayjs"));
var _enhanceDateAdapter = require("./enhanceDateAdapter");
const _dayjsAdapter = {
  // parse: (value: any, format: string) => {
  //   if (value === "") {
  //     return null;
  //   }

  //   return dayjs(value, format, locale, true);
  // },

  date: value => {
    if (value === null) {
      return null;
    }
    return (0, _dayjs.default)(value);
  },
  isValid: value => (0, _dayjs.default)(value).isValid(),
  isAfter: (date, value) => date.isAfter(value),
  isBefore: (date, value) => date.isBefore(value),
  isAfterDay: (date, value) => date.isAfter(value, 'day'),
  isBeforeDay: (date, value) => date.isBefore(value, 'day'),
  isAfterMonth: (date, value) => date.isAfter(value, 'month'),
  isBeforeMonth: (date, value) => date.isBefore(value, 'month'),
  isBeforeYear: (date, value) => date.isBefore(value, 'year'),
  isAfterYear: (date, value) => date.isAfter(value, 'year'),
  formatByString: (date, formatString) => (0, _dayjs.default)(date).format(formatString),
  getHours: date => date.hour(),
  addDays: (date, count) => count < 0 ? date.subtract(Math.abs(count), 'day') : date.add(count, 'day'),
  addMonths: (date, count) => count < 0 ? date.subtract(Math.abs(count), 'month') : date.add(count, 'month'),
  addYears: (date, count) => count < 0 ? date.subtract(Math.abs(count), 'year') : date.add(count, 'year'),
  setMonth: (date, count) => date.set('month', count),
  setHours: (date, count) => date.set('hour', count),
  getMinutes: date => date.minute(),
  setMinutes: (date, count) => date.set('minute', count),
  getSeconds: date => date.second(),
  setSeconds: (date, count) => date.set('second', count),
  getMonth: date => date.month(),
  getDate: date => date.date(),
  setDate: (date, count) => date.set('date', count),
  getDaysInMonth: date => date.daysInMonth(),
  isSameDay: (date, comparing) => date.isSame(comparing, 'day'),
  isSameMonth: (date, comparing) => date.isSame(comparing, 'month'),
  isSameYear: (date, comparing) => date.isSame(comparing, 'year'),
  getYear: date => date.year(),
  setYear: (date, year) => date.set('year', year),
  isEqual: (value, comparing) => {
    if (value === null && comparing === null) {
      return true;
    }
    return (0, _dayjs.default)(value).isSame(comparing);
  },
  formatWithLocale(date1, locale, format) {
    return date1.locale(locale).format(format);
  },
  parseWithLocale(text, locale, format) {
    return (0, _dayjs.default)(text, format, locale, true);
  },
  getDayOfWeek: value => value.day(),
  clone: value => value.clone(),
  startOfYear: date => date.startOf('year'),
  endOfYear: date => date.endOf('year'),
  startOfMonth: date => date.startOf('month'),
  endOfMonth: date => date.endOf('month')
};
/**
 * @public
 */
const dayjsAdapter = exports.dayjsAdapter = /* #__PURE__ */(0, _enhanceDateAdapter.enhanceDateAdapter)(_dayjsAdapter);
