// @universe-design/react-provider/lib-commonjs/date-adapters/momentjsAdapter.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.momentjsAdapter = void 0;
var _moment = _interopRequireDefault(require("moment"));
var _enhanceDateAdapter = require("./enhanceDateAdapter");
const baseMomentjsAdapter = {
  date: value => {
    if (value === null) {
      return null;
    }
    return (0, _moment.default)(value);
  },
  isValid: value => (0, _moment.default)(value).isValid(),
  isAfter: (date, value) => date.isAfter(value),
  isBefore: (date, value) => date.isBefore(value),
  isAfterDay: (date, value) => date.isAfter(value, 'day'),
  isBeforeDay: (date, value) => date.isBefore(value, 'day'),
  isBeforeMonth: (date, value) => date.isBefore(value, 'month'),
  isAfterMonth: (date, value) => date.isAfter(value, 'month'),
  isBeforeYear: (date, value) => date.isBefore(value, 'year'),
  isAfterYear: (date, value) => date.isAfter(value, 'year'),
  formatByString: (date, formatString) => {
    const clonedDate = date.clone();
    return clonedDate.format(formatString);
  },
  getHours: date => date.get('hours'),
  addDays: (date, count) => count < 0 ? date.clone().subtract(Math.abs(count), 'days') : date.clone().add(count, 'days'),
  addMonths: (date, count) => count < 0 ? date.clone().subtract(Math.abs(count), 'months') : date.clone().add(count, 'months'),
  addYears: (date, count) => count < 0 ? date.clone().subtract(Math.abs(count), 'years') : date.clone().add(count, 'years'),
  setHours: (date, count) => date.clone().hours(count),
  getMinutes: date => date.get('minutes'),
  setMinutes: (date, count) => date.clone().minutes(count),
  getSeconds: date => date.get('seconds'),
  setSeconds: (date, count) => date.clone().seconds(count),
  getMonth: date => date.get('month'),
  getDaysInMonth: date => date.daysInMonth(),
  isSameDay: (date, comparing) => date.isSame(comparing, 'day'),
  isSameMonth: (date, comparing) => date.isSame(comparing, 'month'),
  isSameYear: (date, comparing) => date.isSame(comparing, 'year'),
  setMonth: (date, count) => date.clone().month(count),
  startOfYear: date => date.clone().startOf('year'),
  endOfYear: date => date.clone().endOf('year'),
  startOfMonth: date => date.clone().startOf('month'),
  endOfMonth: date => date.clone().endOf('month'),
  getYear: date => date.get('year'),
  setYear: (date, year) => date.clone().set('year', year),
  getDate: date => date.get('date'),
  setDate: (date, year) => date.clone().set('date', year),
  isEqual: (v1, v2) => {
    if (!v1 && !v2) {
      return true;
    }
    if (!(v1 && v2)) {
      return false;
    }
    return (0, _moment.default)(v1).isSame(v2);
  },
  formatWithLocale(date1, locale, format) {
    return date1.locale(locale).format(format);
  },
  parseWithLocale: (text, locale, format) => (0, _moment.default)(text, format, locale, true),
  getDayOfWeek: value => value.day(),
  clone: value => value.clone()
};
/**
 * @public
 */
const momentjsAdapter = exports.momentjsAdapter = /* #__PURE__ */(0, _enhanceDateAdapter.enhanceDateAdapter)(baseMomentjsAdapter);
