// @universe-design/react/src/cascader/cascade/cascadeUI.tsx
/* eslint-disable max-lines-per-function */
/* eslint-disable @typescript-eslint/no-non-null-assertion */
import React, { useMemo } from 'react';
import type { ListRef } from '@universe-design/rc-virtual-list';
import List from '@universe-design/rc-virtual-list';
import type { DoublyTreeNode } from '../../tree';
import { isLeaf } from '../../tree';
import { useCascaderMenuCssScope } from '../common';
import CascaderOption from './Option';
import { HighlightProvider } from '../../shared';
import { useCascadeContext } from './CascadeContext';

export const CascadeUI: React.FC = () => {
  const cssScope = useCascaderMenuCssScope();

  const {
    doublyTreeNodeList: treeNodeList,
    listHeight: listHeightProp,
    dropdownLevelWidth,
    searchValue,
    searchFlattenTree,
    searching,
    expandedList,
    checkedCache,
    selectedCache,
    levelsListRef,
    fullHeight,
    virtual,
  } = useCascadeContext();

  const listHeight = listHeightProp - 2; // 减去 border

  // =========================== List ===========================
  const listRef = React.useRef<ListRef>(null);

  /**
   * ======== Level Dropdown ========
   */

  /**
   * 渲染 level panel
   * @param parentOption level panel 的父级 option
   * @param highlightOption level panel 需要高亮的 option
   */
  const levelRenderer = (
    parentOption: DoublyTreeNode | null,
    highlightOption: DoublyTreeNode | null,
    index: number,
  ) => {
    let options: DoublyTreeNode[] = parentOption?.children || [];
    if (!parentOption) {
      options = treeNodeList.filter(o => !o.parent);
    } else if (isLeaf(parentOption.__node__)) {
      return null;
    }

    return (
      <div
        key={parentOption?.__node__.value || 'UD_CASCADER_RESERVED_ROOT_KEY'}
        className={cssScope('__level')}
        style={{ width: dropdownLevelWidth }}
      >
        <div className={cssScope('__level__list')}>
          <List<DoublyTreeNode>
            itemKey={node => node.__node__.value}
            ref={node => (levelsListRef.current[index] = node)}
            data={options}
            height={listHeight - 2 * 8 - 2 * 1}
            itemHeight={32}
            fullHeight={fullHeight}
            // onScroll={onPopupScroll}
            virtual={virtual({ type: 'levelMenu', context: { parentOption, level: index + 1 } })}
            // onMouseEnter={onMouseEnter}
          >
            {childNode => {
              const {
                __node__: { value: key },
              } = childNode;
              return (
                // 这里加一个空 div 是为了把内部带 margin 的元素包住，让 virtual list 能正常计算
                <div>
                  <CascaderOption
                    isChecked={checkedCache.get(key)}
                    isSelected={selectedCache.get(key)}
                    showExpandIcon={true}
                    key={key}
                    node={childNode}
                    highlightOption={highlightOption}
                  />
                </div>
              );
            }}
          </List>
        </div>
      </div>
    );
  };

  let menuNode: React.ReactElement;

  if (searching) {
    menuNode = (
      <div className={cssScope('__list')}>
        <List<DoublyTreeNode>
          itemKey={node => node.__node__.value}
          ref={listRef}
          data={searchFlattenTree}
          height={listHeight - 2 * 3 - 2 * 1}
          itemHeight={32}
          fullHeight={false}
          // onScroll={onPopupScroll}
          virtual={virtual({ type: 'searchMenu', context: undefined })}
          // onMouseEnter={onMouseEnter}
        >
          {childNode => {
            const {
              __node__: { value: key },
            } = childNode;
            return (
              // 这里加一个空 div 是为了把内部带 margin 的元素包住，让 virtual list 能正常计算，tree，select，cascader
              <div>
                <CascaderOption
                  isChecked={checkedCache.get(key)}
                  isSelected={selectedCache.get(key)}
                  showExpandIcon={false}
                  key={key}
                  node={childNode}
                  highlightOption={null}
                />
              </div>
            );
          }}
        </List>
      </div>
    );
  } else {
    menuNode = (
      <div className={cssScope('__level__wrapper')}>
        {expandedList.map(({ parent, highlight }, index) => levelRenderer(parent, highlight, index))}
      </div>
    );
  }

  return (
    <HighlightProvider
      value={useMemo(
        () => ({
          searchWords: searchValue ? [searchValue] : undefined,
          disabled: false,
        }),
        [searchValue],
      )}
    >
      {menuNode}
    </HighlightProvider>
  );
};
