// @universe-design/react/src/cascader/cascade/interface.ts
import type React from 'react';
import type { DoublyTreeNode, TreeNodeConfig, TreeProps, TreeCheckProps, ShowCheckedStrategy } from '../../tree';
import type { ListScrollConfig } from '../../util';
import type { RefSelectRef } from '../../select';

/**
 * @public
 */
export interface CascaderFilterOptionFunc {
  (inputValue: string, node: DoublyTreeNode): boolean;
}

/**
 * @public
 */
export type CascaderVirtualParams =
  | {
      type: 'levelMenu';
      context: {
        parentOption: DoublyTreeNode | null;
        level: number;
      };
    }
  | {
      type: 'searchMenu';
      context: undefined;
    };

/**
 * @public
 * 用于在 Cascader 聚合使用
 */
export interface CascadeSharedProps extends Pick<TreeProps, 'loadData' | 'loadedKeys' | 'onLoad'> {
  /**
   * 下拉框的宽度
   * @defaultValue 160
   */
  dropdownLevelWidth?: number;
  /**
   * 下拉框的高度
   * @defaultValue 256
   */
  listHeight?: number;
  /**
   * 过滤选项的筛选函数
   * @defaultValue defaultFilterOption
   */
  filterOption?: CascaderFilterOptionFunc;
  /**
   * 筛选时对结果的排序函数, 类似 [Array.sort](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort) 里的 `compareFunction`
   */
  filterResultSorter?: (path1: TreeNodeConfig[], path2: TreeNodeConfig[], inputValue: string) => number;
  /**
   * 筛选时最多显示多少条结果，如果设置为 false 则不限制数量
   * @defaultValue 50
   */
  filterResultLimit?: number | false;
  /**
   * 自定义筛选结果的渲染函数
   */
  filterResultRenderer?: (path: TreeNodeConfig[], inputValue: string) => React.ReactNode;
  /**
   * 次级菜单的展开方式
   * @defaultValue `click`
   */
  expandTrigger?: 'click' | 'hover';
  /**
   * 单选的时候是否可以选中非叶子节点，多选时默认开启，单选时默认不开启
   */
  changeOnSelect?: boolean;
  /**
   * 定义选中项回填的方式。 `Cascader.SHOW_PARENT`：当父节点下所有子节点都选中时，只显示父节点；`Cascader.SHOW_CHILD`：当父节点下所有子节点都选中时，只显示子节点。
   * @defaultValue Cascader.SHOW_CHILD
   */
  showCheckedStrategy?: ShowCheckedStrategy.SHOW_CHILD | ShowCheckedStrategy.SHOW_PARENT;
  /**
   * 是否是可勾选。只能进行多选，并且默认遵从 `仅显示子级` 策略，如果父级的所有子级都为选中状态，则只显示子级节点的值
   * @defaultValue false
   */
  checkable?: boolean;
  /**
   * 下拉菜单是否占满最大高度
   * @defaultValue true
   */
  fullHeight?: boolean;
  /**
   * 下拉菜单可逐级判断是否开启虚拟滚动。
   * @defaultValue () => true
   */
  virtual?: (params: CascaderVirtualParams) => boolean;
}

/**
 * @internal
 */
export type CascadeProps = CascadeSharedProps &
  Pick<TreeProps, 'onSelect' | 'multiple' | 'searchValue'> &
  Pick<TreeCheckProps, 'onCheck'> & {
    /**
     * 选项
     */
    options?: TreeNodeConfig[];
    /**
     * 选中值
     */
    value?: string[];
    /**
     * 选中值默认值
     */
    defaultValue?: string[];
    /**
     * 选中值变化时的回调
     */
    onChange?: (value: string[]) => void;
  };

export interface CascaderRef extends RefSelectRef {
  /**
   * 将指定选项滚动到可视区域内
   */
  scrollIntoView: (key: string, scrollConfig?: ListScrollConfig) => void;
}
