// @universe-design/react/src/react-mini-form/internal/FormInternalContext.tsx
import type { SetStateAction } from 'react';
import type React from 'react';
import { createContext, useContext, useContextSelector } from 'use-context-selector';
import type { DeepPartial } from 'utility-types';
import type { Subject } from 'rxjs';
import type {
  FormErrors,
  FormTouched,
  Path,
  FinalValidator,
  ValidateOptions,
  ValidateEventParam,
  SerializedName,
} from '../shared';
import { createContextHook } from '@universe-design/react-utilities';

export interface RegistryValue {
  fieldID: string;
  serializedName: SerializedName; // 序列化后的 namePath
  validate: React.MutableRefObject<FinalValidator[]>;
}

export interface SetFieldValueConfig {
  shouldValidate?: boolean;
  triggerByUser?: boolean;
  triggerName?: string;
}

/**
 * @docgen
 * @typeParam Values - values of this form
 */
export interface FormInternalContextType<Values> {
  /**
   * 所有字段的值，此值为最新值，谨慎在异步回调中使用
   */
  readonly values: Values;
  /**
   * 所有字段的错误信息。
   */
  readonly errors: FormErrors<Values> | undefined;
  /**
   * 所有字段是否被修改过。
   */
  readonly touched: FormTouched<Values>;
  /**
   * 是否在提交过程。
   */
  isSubmitting: boolean;
  /**
   * 是否在校验过程。
   */
  isValidating: boolean;
  /**
   * 是否有错误存在
   */
  isDirty: boolean;
  /**
   *  validate订阅中心，用于校验字段
   * @internal
   */
  validate$: Subject<ValidateEventParam>;
  /**
   * 覆盖设置多个表单项的值，全量更新。
   */
  setValues: (values: SetStateAction<Values>, shouldValidate?: boolean) => void;
  /**
   * 设置多个表单项的值，传入对象属于一个表单项的部分进行替换，否则递归进行增量更新
   */
  setFieldsValue: (
    values: DeepPartial<Values> | ((values: Values) => DeepPartial<Values>),
    shouldValidate?: boolean,
  ) => void;
  /**
   * 设置一个表单项的值，对 Path 位置全量更新
   */
  setFieldValue: <S>(name: Path, value: S | ((values: Values) => S), shouldValidate?: boolean) => void;
  _setFieldValue: <S>(
    serializedName: SerializedName,
    value: S | ((values: Values) => S),
    config: SetFieldValueConfig,
  ) => void;
  _getStateValues: () => Values;
  /**
   * 覆盖设置多个表单项的错误，全量更新。
   */
  setErrors: (errors: SetStateAction<FormErrors<Values> | undefined>) => void;
  /**
   * 重置所有表单项的错误。
   */
  resetErrors: () => void;
  /**
   * 设置多个表单项的错误，传入对象属于一个表单项的部分进行替换，否则递归进行增量更新
   */
  setFieldsError: (
    values:
      | DeepPartial<FormErrors<Values>>
      | ((errors: FormErrors<Values> | undefined) => DeepPartial<FormErrors<Values>>),
  ) => void;
  /**
   * 设置一个表单项的错误。
   */
  setFieldError: <S>(name: Path, error: S | ((errors: FormErrors<Values> | undefined) => S)) => void;
  _setFieldError: <S>(
    serializedName: SerializedName,
    error: S | ((errors: FormErrors<Values> | undefined) => S),
  ) => void;
  /**
   * 手动提交表单。
   */
  submit: () => Promise<void>;
  /**
   * 重置表单到初始值。
   */
  reset: () => void;
  /**
   * 手动校验字段，可指定校验的字段以及触发校验的事件
   */
  validateFields: (nameList?: Path[], options?: ValidateOptions) => Promise<FormErrors<Values> | void>;
  /**
   * 设置一个字段是否被修改过。
   * @internal
   */
  _setFieldTouched: (serializedName: SerializedName, touched: boolean) => void;
  _setFieldTouchedState: (serializedName: SerializedName, touchedState: Record<any, any>) => void;
  /**
   * 把 field 注册到 form 中去，validate 的时候遍历使用。
   * @internal
   */
  _registerField: (payload: RegistryValue) => void;
  /**
   * 从 from 中注销 field。
   * @internal
   */
  _unregisterField: (fieldID: string) => void;
}

const {
  Provider: FormInternalContextProvider,
  useContext: _useFormInternalContext,
  useContextSelector: _useFormInternalContextSelector,
} = createContextHook<FormInternalContextType<any>>('MiniFormContext', {
  strict: true,
  customizedContextMethods: {
    createContext,
    useContext,
    useContextSelector,
  },
});

export { FormInternalContextProvider };

/**
 * useFormInternalContextSelector
 * get form context, similar to redux selector
 */
export function useFormInternalContextSelector<Values, Selected>(
  selector: (value: FormInternalContextType<Values>) => Selected,
): Selected {
  return _useFormInternalContextSelector(selector);
}

/**
 * useFormInternalContext
 * get from context
 */
export function useFormInternalContext<Values>() {
  return _useFormInternalContext() as FormInternalContextType<Values>;
}
