// @universe-design/react/src/react-mini-form/internal/useFieldInternal.tsx
import get from 'lodash/get';
import * as React from 'react';
import { useConstant, useForceUpdate } from '@universe-design/react-utilities';
import { useLatest } from '../../hooks';
import type { LiteralUnion } from '../../util';
import type { ValidateOptions, ValidateTrigger, SerializedName, Path, Validator } from '../shared';
import { marshalNamePath, asArray, normalizeValidator, DEFAULT_VALIDATE_TRIGGER } from '../shared';
import { useFormInternalContextSelector } from './FormInternalContext';

/**
 * @public
 * @docgen
 */
export type UseFieldReturnValue = Pick<UseFieldInternalReturnValue, 'value' | 'handleChange' | 'error'>;

/**
 * @public
 */
export type FieldProps = {
  /**
   * Field children, can either be a `ReactElement` or a `RenderProp`。
   */
  children: React.ReactElement | ((props: UseFieldReturnValue) => React.ReactNode);
  /**
   * 表单项的名称。
   */
  name: Path;
  /**
   * 子节点的值属性, 仅 `ReactElement` 子节点可用。
   */
  valuePropName?: LiteralUnion<'value' | 'checked', string>;
  /**
   * 设置收集字段值变更的回调函数名称, 仅 `ReactElement` 子节点可用。
   * @defaultValue `onChange`
   */
  trigger?: string;
  /**
   * 设置字段校验的回调函数名称, 仅 `ReactElement` 子节点可用。
   */
  validateTrigger?: string | string[];
  /**
   * 表单级别的校验函数，返回一个 `string` 作为错误信息，如果返回空则认为校验通过，该函数可以是一个异步函数，并且可以配置`validateTrigger`。
   */
  validate?: Validator | Validator[];
  /**
   * 校验状态，如不设置，则会根据校验规则自动生成，可选 `error`。
   */
  validateStatus?: 'error';
};

/**
 * @public
 * @docgen
 */
export type UseFieldProps = Pick<FieldProps, 'name' | 'validate' | 'valuePropName' | 'validateTrigger' | 'trigger'>;

/**
 * useFieldContext
 */
export function useFieldContext({ name }: Partial<Pick<FieldProps, 'name'>>) {
  const multiLevelPath = asArray(name);
  return {
    value: useFormInternalContextSelector(v => get(v.values, multiLevelPath)),
    error: useFormInternalContextSelector(v => (v.errors ? get(v.errors, multiLevelPath) : undefined)),
  };
}

export interface UseFieldInternalReturnValue {
  /**
   * 字段受控值。
   */
  value: any;
  /**
   * 字段变更处理事件，第一个参数是可以是 `event` 或者是确切的字段值。
   */
  handleChange: (eventOrValue: any) => void;
  /**
   * 字段的错误信息。
   */
  error?: string;
  allValidateTriggers: ValidateTrigger[];
  serializedName: SerializedName;
  trigger: string;
  validateField: (options?: ValidateOptions) => void;
}

let count = 0;
const useFieldID = () => useConstant(() => `field-${count++}`);

/**
 * useField
 */
export function useFieldInternal(props: UseFieldProps): UseFieldInternalReturnValue {
  const { name, validate, valuePropName, validateTrigger, trigger = 'onChange' } = props;
  const fieldID = useFieldID();
  const namePathStr = marshalNamePath(name);
  const { value, error } = useFieldContext({ name });
  const _setFieldValue = useFormInternalContextSelector(v => v._setFieldValue);
  const _registerField = useFormInternalContextSelector(v => v._registerField);
  const _unregisterField = useFormInternalContextSelector(v => v._unregisterField);
  const _setFieldTouched = useFormInternalContextSelector(v => v._setFieldTouched);
  const validate$ = useFormInternalContextSelector(v => v.validate$);

  const forceUpdate = useForceUpdate();

  // validate 统一为 Validator[]
  // 1、当 Validator 没有设置 validateTrigger，会使用 Field 的 validateTrigger
  // 2、在 1 条件下，当 field 没有设置validateTrigger，会使用默认的 validateTrigger
  const mergedValidators = (() => {
    const validators = asArray(validate).map(normalizeValidator);
    validators.forEach(v => {
      if (!v.validateTrigger.length) {
        v.validateTrigger = asArray(validateTrigger);
      }
      if (!v.validateTrigger.length) {
        v.validateTrigger.push(DEFAULT_VALIDATE_TRIGGER);
      }
    });
    return validators;
  })();

  const mergedValidatorsRef = useLatest(mergedValidators);

  const allValidateTriggers = (() => {
    const triggerSet = new Set<string>();
    mergedValidators.forEach(v => {
      v.validateTrigger.forEach(t => {
        if (t !== DEFAULT_VALIDATE_TRIGGER) {
          triggerSet.add(t);
        }
      });
    });
    return [...triggerSet];
  })();

  React.useEffect(() => {
    _registerField({
      fieldID,
      serializedName: namePathStr,
      validate: mergedValidatorsRef,
    });
    return () => {
      _unregisterField(fieldID);
    };
  }, [_registerField, _unregisterField, mergedValidatorsRef, fieldID, namePathStr]);

  /**
   * Super mega handle Change
   */
  const handleChangeMega = React.useCallback(
    (eventOrValue: any) => {
      /**
       * 此处 forceUpdate 是因为用了 use-context-selector 导致 change 之后 Field 没有重新渲染，
       * 导致 child 拿到旧值，引发光标位置的抖动
       * issue： https://code.byted.org/universe-ui/react/issues/937
       */
      forceUpdate();
      // First arg is event
      let _value: any;

      if (
        typeof eventOrValue === 'object' &&
        'nativeEvent' in eventOrValue &&
        eventOrValue.nativeEvent instanceof Event
      ) {
        if (eventOrValue.target instanceof HTMLInputElement) {
          switch (eventOrValue.target.type) {
            case 'radio':
            case 'checkbox':
              _value = eventOrValue.target[valuePropName ?? 'checked'];
              break;
            case 'file':
              throw new Error(
                'Field cannot be used on file inputs since they are read-only. Use a onChange listener instead.',
              );
            default:
              _value = eventOrValue.target[valuePropName ?? 'value'];
          }
        } else if (eventOrValue.target instanceof HTMLSelectElement) {
          if (!eventOrValue.target.multiple) {
            _value = eventOrValue.target.value;
          } else {
            const res: string[] = [];
            for (let i = 0; i < eventOrValue.target.options.length; i++) {
              const opt = eventOrValue.target.options[i];
              if (opt.selected) {
                res.push(opt.value);
              }
            }
            _value = res;
          }
        } else {
          // radio group
          switch (eventOrValue.target.type) {
            case 'radio':
            case 'checkbox':
              _value = eventOrValue.target[valuePropName ?? 'checked'];
              break;
            default:
              _value = eventOrValue.target[valuePropName ?? 'value'];
              break;
          }
        }
      } else {
        _value = eventOrValue;
      }
      _setFieldTouched(namePathStr, true);
      _setFieldValue(namePathStr, _value, {
        triggerByUser: true,
        triggerName: trigger,
        shouldValidate: true,
      });
    },
    [_setFieldTouched, _setFieldValue, forceUpdate, namePathStr, trigger, valuePropName],
  );

  const validateField = React.useCallback(
    (options: ValidateOptions | undefined) => {
      validate$.next({
        serializedName: namePathStr,
        triggerName: options?.triggerName || DEFAULT_VALIDATE_TRIGGER,
      });
    },
    [namePathStr, validate$],
  );

  return {
    value,
    handleChange: handleChangeMega,
    error,
    allValidateTriggers,
    serializedName: namePathStr,
    trigger,
    validateField,
  };
}
