// @universe-design/react/src/react-mini-form/internal/useFormInternal.tsx
/* eslint-disable max-depth */
// nolint: long_function
import get from 'lodash/get';
import set from 'lodash/set';
import pick from 'lodash/pick';
import type React from 'react';
import { useCallback, useEffect, useRef, useState } from 'react';
import type { Observable } from 'rxjs';
import { Subject, from, forkJoin, of, throwError, firstValueFrom } from 'rxjs';
import { switchMap, map, tap, delay } from 'rxjs/operators';
import { useConstant, useLogger } from '@universe-design/react-utilities';
import { useEffectOnce, useEventCallback, useMountedState } from '../../hooks';
import { isFunction, isPromise, __PROD__ } from '../../util';
import type { FormErrors, Path, FormTouched, ValidateEventParam, SerializedName } from '../shared';
import {
  marshalNamePath,
  mergeWithUndefined,
  printFormValidationError,
  unmarshalNamePath,
  DEFAULT_VALIDATE_TRIGGER,
} from '../shared';
import { produce, freeze } from './formImmer';
import type { UseFormProps } from '../Form';
import type { FormInternalContextType, RegistryValue } from './FormInternalContext';

function resolveStateActionsValue<PrevValues, ActionsValue>(
  prevValues: PrevValues,
  actionsValue: ActionsValue,
): ActionsValue extends (...args: any[]) => any ? ReturnType<ActionsValue> : ActionsValue {
  const resolvedValues = isFunction(actionsValue) ? actionsValue(prevValues) : actionsValue;

  return resolvedValues;
}

/**
 * Form State: 1. values; 2. errors; 3. toucheds
 * FromItem change 触发 setValues 并且触发 validate
 *
 * validate 包含两个个环节: 1. 所有表单项级别 validate; 2. 表单级别 validate
 * 表单项 validate 只对已经 touched 和已经注册的字段进行校验
 */
export function useForm<Values extends Record<any, any>>(props: UseFormProps<Values>): FormInternalContextType<Values> {
  /**
     * FIXME use event callback may cause problems in some rare cases. Assume that you have a form and a single input, now you want to do some check for the input value and form value in `onSubmit` callback. Assume that you have a long-time async validation, and you change the single input value to `abc` from empty during validation. If use useCallback, the input value in  `onSubmit`'s closure will be `""`, but when using useEventCallback, the input value will be `"abc"`
     * @example
     * ```
      function delayReturn<T>(value: T, timeout?: number) {
        return new Promise<T>((resolve) => {
          window.setTimeout(() => {
            resolve(value);
          }, timeout);
        });
      }

      function A() {
        const [value, setValue] = useState<string>("");
        return (
          <>
            <input value={value} onChange={(e) => setValue(e.target.value)} />
            <Form
              initialValues={initialValues}
              onSubmit={() => {
                alert(value);
              }}
            >
              <input type={"submit"} />
              <Field name={["input"]} validate={() => delayReturn(void 0, 5000)}>
                <input />
              </Field>
            </Form>
          </>
        );
      }
     * ```
     */
  const onSubmit = useEventCallback(props.onSubmit);
  const validate = useEventCallback(props.validate);
  const onValuesChange = useEventCallback(props.onValuesChange);
  const onSubmitError = useEventCallback(props.onSubmitError);
  const { initialValues, debug = false } = props;

  // ================ From Context ================

  // form 注册中心，用于记录 form item 的 name validator 等，使用 field id 作为 key
  const registry = useConstant(() => new Map<string, RegistryValue>());
  const serializedNameRegistryValueIndexes = useConstant(() => new Map<SerializedName, RegistryValue>());

  // values
  const [values, _setStateValues] = useState<Values>(freeze(initialValues));
  const valuesRef = useRef(values);
  const setStateValues = useCallback(
    (
      nextValues: Values | ((prevValues: Values) => Values),
      changedConfig?: {
        triggerByUser: boolean | undefined;
        changedNamePath: string;
      },
    ) => {
      const nextV: Values = freeze(resolveStateActionsValue(valuesRef.current, nextValues));
      valuesRef.current = nextV;
      _setStateValues(nextV);
      if (changedConfig?.triggerByUser && isFunction(onValuesChange)) {
        onValuesChange(pick(nextV, [unmarshalNamePath(changedConfig.changedNamePath)]) as Values, nextV);
      }
    },
    [onValuesChange],
  );

  // errors
  const [errors, _setStateErrors] = useState<FormErrors<Values>>();
  const errorsRef = useRef(errors);
  const setStateErrors = useCallback<React.Dispatch<React.SetStateAction<FormErrors<Values> | undefined>>>(
    nextValues => {
      const nextV: FormErrors<Values> | undefined = freeze(resolveStateActionsValue(errorsRef.current, nextValues));
      errorsRef.current = nextV;
      _setStateErrors(nextV);
    },
    [],
  );

  // touched
  // latest value store，not a state store
  const touchedRef = useRef({} as FormTouched<Values>);

  const _setFieldTouched = useCallback<FormInternalContextType<Values>['_setFieldTouched']>(
    (serializedName, _touched) => {
      set(touchedRef.current, unmarshalNamePath(serializedName), _touched);
    },
    [],
  );

  const _setFieldTouchedState = useCallback<FormInternalContextType<Values>['_setFieldTouchedState']>(
    (serializedName, _touched) => {
      set(touchedRef.current, unmarshalNamePath(serializedName), _touched);
    },
    [],
  );

  debug &&
    !__PROD__ &&
    // eslint-disable-next-line react-hooks/rules-of-hooks
    useLogger('react-mini-form', {
      registry,
      values,
      errors,
      touched: touchedRef.current,
    });

  /**
   * ================ Validation ================
   */

  // isValidating
  const [isValidating, setIsValidating] = useState(false);
  // validate 订阅中心
  const validate$ = useConstant(() => new Subject<ValidateEventParam>());

  const _validateSingleFieldLevel$ = useCallback(
    function _validateSingleFieldLevel$(
      serializedName: SerializedName,
      event: ValidateEventParam,
    ): Observable<{
      errorMessage: string | void;
      skipped: boolean;
    }> {
      const validators = serializedNameRegistryValueIndexes.get(serializedName)?.validate.current ?? [];
      const filteredValidators = validators.filter(
        v =>
          // if name is not undefined, this is triggered by onChange, so we should only validate this items
          (event.serializedName === undefined ||
            (event.serializedName === serializedName && get(touchedRef.current, unmarshalNamePath(serializedName)))) &&
          // namePaths from validateFields contains current field
          (event.namePaths === undefined || event.namePaths.some(p => marshalNamePath(p) === serializedName)) &&
          // current field's validateTrigger contains trigger event
          (v.validateTrigger.includes(DEFAULT_VALIDATE_TRIGGER) ||
            event.triggerName === DEFAULT_VALIDATE_TRIGGER ||
            v.validateTrigger.includes(event.triggerName)),
      );

      // 如果没有符合条件的 validators 就标记为跳过，后面就不处理 combined errors 逻辑了
      if (!filteredValidators.length) {
        return of({
          errorMessage: undefined,
          skipped: true,
        });
      }

      const validatePromise = new Promise<{
        errorMessage: string | void;
        skipped: boolean;
      }>(async resolve => {
        for (const validator of filteredValidators) {
          const maybeErrorMessage = await validator.validate(event.value, unmarshalNamePath(serializedName));
          if (maybeErrorMessage) {
            resolve({
              errorMessage: maybeErrorMessage,
              skipped: false,
            });
            return;
          }
        }
        resolve({
          errorMessage: undefined,
          skipped: false,
        });
      });
      return from(validatePromise);
    },
    [serializedNameRegistryValueIndexes],
  );

  // mounted
  const isMounted = useMountedState();

  const _validateMultipleFieldLevel$ = useCallback(
    (event: ValidateEventParam): Observable<FormErrors<Values>> =>
      // TODO TO BE CHECKED!! Please pay attention, we will do all field level validation for now.
      // if (name) {
      //   const k = marshalNamePath(name);
      //   const v = registry.get(k);
      //   if (v) {
      //     nameEntries.push([k, v]);
      //   }
      // } else {
      //   nameEntries = [...registry.entries()];
      // }
      // console.log('_validateMultipleFieldLevel$', nameEntries);
      forkJoin(
        // TODO: field 的每个 error 互不关联，不需要等待全部返回
        [...registry.values()].map(v => {
          const value = get(valuesRef.current, unmarshalNamePath(v.serializedName));
          return _validateSingleFieldLevel$(v.serializedName, {
            value,
            ...event,
          });
        }),
      ).pipe(
        map(_errors =>
          [...registry.values()].reduce((pre, v, i) => {
            const error = _errors[i];
            const name = unmarshalNamePath(v.serializedName);
            if (error.skipped) {
              set(pre, name, get(errorsRef.current, name));
            } else {
              set(pre, name, error.errorMessage);
            }
            return pre;
          }, {} as unknown as FormErrors<Values>),
        ),
      ),
    [_validateSingleFieldLevel$, registry],
  );

  /**
   * _validateMultipleFieldLevel
   * @param values current values state
   * @param name if undefined, validate all fields, if assigned, validate name.
   */
  const _validateFormLevel$ = useCallback(
    (name?: Path): Observable<FormErrors<Values> | void> => {
      try {
        if (!isFunction(validate)) {
          return of(undefined);
        }
        const maybePromise = validate(valuesRef.current, name);
        if (!maybePromise) {
          return of(undefined);
        }
        if (isPromise(maybePromise)) {
          return from(maybePromise);
        }
        return of(maybePromise);
      } catch (err) {
        // sync error to observable
        return throwError(err);
      }
    },
    [validate],
  );

  const _validateAll$ = useCallback(
    (event: ValidateEventParam): Observable<FormErrors<Values>> =>
      forkJoin([_validateMultipleFieldLevel$(event), _validateFormLevel$(event.serializedName)]).pipe(
        map(([fieldErrors, formErrors]) => {
          // 这里的 merge 要用 lodash merge，因为我们不能让 formErrors 的 undefined 覆盖 fieldErrors 里面已有的 errors
          const combinedErrors = {} as FormErrors<Values>;
          registry.forEach(({ serializedName: _serializedName }) => {
            const name = unmarshalNamePath(_serializedName);
            const fieldErrorMessage = get(fieldErrors, name);
            const formErrorMessage = get(formErrors, name);
            const errorMessage = fieldErrorMessage || formErrorMessage;
            if (errorMessage) {
              // 如果出现了 error 则认为是 touch 了
              _setFieldTouched(_serializedName, true);
              set(combinedErrors, name, errorMessage);
            }
          });
          return combinedErrors;
        }),
      ),
    [_validateMultipleFieldLevel$, _validateFormLevel$, registry, _setFieldTouched],
  );

  useEffect(() => {
    const sub = validate$
      .pipe(
        tap(() => {
          setIsValidating(true);
        }),
        delay(0),
        switchMap(({ serializedName, triggerName }) => _validateAll$({ serializedName, triggerName })),
      )
      .subscribe({
        next: combinedErrors => {
          if (isMounted()) {
            setIsValidating(false);
            setStateErrors(combinedErrors);
          }
        },
        error: error => {
          printFormValidationError(error);
          if (isMounted()) {
            setIsValidating(false);
          }
        },
        complete: () => {
          if (isMounted()) {
            setIsValidating(false);
          }
        },
      });
    return () => {
      sub.unsubscribe();
    };
  }, [_validateAll$, isMounted, setStateErrors, validate$, valuesRef]);

  useEffectOnce(() => () => {
    validate$.complete();
  });

  // TODO rare case
  // 1. change 触发的 validate
  // 2. 修改 value
  // 3. validateFields
  // 4. validateFields 先返回值，change 触发的后返回值
  // 5. 则 errors 有可能会不正确

  type CurrentFormContextType = FormInternalContextType<Values>;

  const validateFields = useCallback<CurrentFormContextType['validateFields']>(
    (nameList, options) => {
      setIsValidating(true);
      return new Promise((resolve, reject) => {
        // 强行在下一次渲染之后进行校验，解决 form item 根据 values 生成场景校验在 form item 注销之前带来的问题。
        setTimeout(() => {
          firstValueFrom(
            _validateAll$({
              namePaths: nameList,
              triggerName: options?.triggerName || DEFAULT_VALIDATE_TRIGGER,
            }),
            { defaultValue: undefined },
          )
            .then(combinedErrors => {
              if (isMounted()) {
                setIsValidating(false);
                setStateErrors(combinedErrors);
              }
              resolve(combinedErrors);
            })
            .catch(err => {
              printFormValidationError(err);
              if (isMounted()) {
                setIsValidating(false);
              }
              reject(err);
            });
        });
      });
    },
    [_validateAll$, isMounted, setStateErrors],
  );

  const setValues = useCallback<CurrentFormContextType['setValues']>(
    (_values, shouldValidate) => {
      setStateValues(prevValues => {
        const resolvedValues = isFunction(_values) ? _values(prevValues) : _values;
        const willValidate = shouldValidate === undefined ? true : shouldValidate;
        if (willValidate) {
          validate$.next({ triggerName: DEFAULT_VALIDATE_TRIGGER });
        }
        return resolvedValues;
      });
    },
    [setStateValues, validate$],
  );

  const setFieldsValue = useCallback<CurrentFormContextType['setFieldsValue']>(
    (partialValues, shouldValidate) => {
      setStateValues(prevValues => {
        const fieldPathSet = new Set<SerializedName>([...registry.values()].map(o => o.serializedName));
        // 这里的 merge 要用 mergeWithUndefined，因为用户可能会把某个 field 设置为 undefined，这个时候用户的预期行为其实是要 merge undefined 值的
        const resolvedValues = produce(prevValues, draft => {
          mergeWithUndefined(
            draft,
            isFunction(partialValues) ? partialValues(prevValues) : partialValues,
            fieldPathSet,
          );
        });
        const willValidate = shouldValidate === undefined ? true : shouldValidate;
        if (willValidate) {
          validate$.next({ triggerName: DEFAULT_VALIDATE_TRIGGER });
        }
        return resolvedValues;
      });
    },
    [registry, setStateValues, validate$],
  );

  const _setFieldValue = useCallback<CurrentFormContextType['_setFieldValue']>(
    (serializedName, value, config) => {
      setStateValues(
        prevValues => {
          const nextValues = produce(prevValues, draft => {
            set(draft, unmarshalNamePath(serializedName), isFunction(value) ? value(prevValues) : value);
          });
          // 若为用户事件触发 onChange，校验逻辑已经 merge 进 trigger 中
          const willValidate = config.shouldValidate === undefined ? !config.triggerByUser : config.shouldValidate;
          if (willValidate) {
            validate$.next({
              serializedName,
              triggerName: config.triggerName || DEFAULT_VALIDATE_TRIGGER,
            });
          }
          return nextValues;
        },
        { triggerByUser: config.triggerByUser, changedNamePath: serializedName },
      );
    },
    [setStateValues, validate$],
  );

  const setFieldValue = useCallback<CurrentFormContextType['setFieldValue']>(
    (name, value, shouldValidate) => _setFieldValue(marshalNamePath(name), value, { shouldValidate }),
    [_setFieldValue],
  );

  const _getStateValues = useCallback<CurrentFormContextType['_getStateValues']>(() => values, [values]);

  const setErrors = useCallback<CurrentFormContextType['setErrors']>(
    _errors => {
      setStateErrors(prevErrors => resolveStateActionsValue(prevErrors, _errors));
    },
    [setStateErrors],
  );

  const resetErrors = useCallback<CurrentFormContextType['resetErrors']>(() => setErrors(undefined), [setErrors]);

  const setFieldsError = useCallback<CurrentFormContextType['setFieldsError']>(
    partialErrors => {
      setStateErrors((prevErrors = {} as FormErrors<Values>) => {
        const fieldPathSet = new Set<SerializedName>([...registry.values()].map(o => o.serializedName));
        // 这里的 merge 要用 mergeWithUndefined，因为用户可能会把某个 field 设置为 undefined，这个时候用户的预期行为其实是要 merge undefined 值的
        return produce(prevErrors, draft => {
          mergeWithUndefined(draft, resolveStateActionsValue(prevErrors, partialErrors), fieldPathSet);
        });
      });
    },
    [registry, setStateErrors],
  );

  const _setFieldError = useCallback<CurrentFormContextType['_setFieldError']>(
    (name, error) => {
      setStateErrors((prevErrors = {} as FormErrors<Values>) =>
        produce(prevErrors, draft => {
          set(draft, unmarshalNamePath(name), resolveStateActionsValue(prevErrors, error));
        }),
      );
    },
    [setStateErrors],
  );

  const setFieldError = useCallback<CurrentFormContextType['setFieldError']>(
    (name, error) => _setFieldError(marshalNamePath(name), error),
    [_setFieldError],
  );

  // ================ Submit ================
  const [isSubmitting, setSubmitting] = useState(false);

  const checkIsDirty = useCallback(
    (_errors: FormErrors<Values> | void) =>
      Boolean(_errors) &&
      [...registry.values()].some(({ serializedName }) => get(_errors, unmarshalNamePath(serializedName))),
    [registry],
  );

  const submit = useCallback<FormInternalContextType<Values>['submit']>(async () => {
    try {
      setSubmitting(true);
      const combinedErrors = await validateFields();
      const isDirty = checkIsDirty(combinedErrors);
      if (isDirty && combinedErrors) {
        isFunction(onSubmitError) && onSubmitError(combinedErrors);
      } else {
        isFunction(onSubmit) && (await onSubmit(values));
      }
      if (isMounted()) {
        setSubmitting(false);
      }
    } catch (err) {
      if (isMounted()) {
        setSubmitting(false);
      }
      throw err;
    }
  }, [validateFields, checkIsDirty, isMounted, onSubmitError, onSubmit, values]);

  const reset = useCallback<FormInternalContextType<Values>['reset']>(() => {
    setStateValues(initialValues);
    setStateErrors({} as FormErrors<Values>);
    touchedRef.current = {} as FormTouched<Values>;
  }, [initialValues, setStateErrors, setStateValues]);

  const _registerField = useCallback<FormInternalContextType<Values>['_registerField']>(
    payload => {
      const { serializedName } = payload;
      if (serializedNameRegistryValueIndexes.get(serializedName)) {
        console.warn(`name: \`${serializedName}\` is duplicated. `);
      }
      registry.set(payload.fieldID, payload);
      serializedNameRegistryValueIndexes.set(serializedName, payload);
    },
    [registry, serializedNameRegistryValueIndexes],
  );

  const _unregisterField = useCallback<FormInternalContextType<Values>['_unregisterField']>(
    fieldID => {
      const registryValue = registry.get(fieldID);
      if (!registryValue) {
        return;
      }
      registry.delete(fieldID);
      /**
       * 「兼容逻辑」
       * 只对 fieldID 和 index registryValue 对得上的情况做 indexes 的清理，因为有可能存在一种情况
       * 首次渲染: list 中只有 index0 的元素
       * 重渲染: list 在 index0 处增加一个新节点，index0 节点的 register 可能会比 index1(之前的 index0) 节点的 unregister 早调用，导致 index0 的 register value 被删除
       */
      if (serializedNameRegistryValueIndexes.get(registryValue.serializedName)?.fieldID === fieldID) {
        serializedNameRegistryValueIndexes.delete(registryValue.serializedName);
      }
      _setFieldTouched(registryValue.serializedName, false);
    },
    [_setFieldTouched, registry, serializedNameRegistryValueIndexes],
  );

  const formCtx: FormInternalContextType<Values> = {
    _getStateValues,
    get values() {
      return valuesRef.current;
    },
    get touched() {
      return touchedRef.current;
    },
    setValues,
    setFieldsValue,
    _setFieldValue,
    setFieldValue,
    _setFieldError,
    setErrors,
    resetErrors,
    setFieldsError,
    setFieldError,
    submit,
    reset,
    get errors() {
      return errorsRef.current;
    },
    isSubmitting,
    isValidating,
    get isDirty() {
      return checkIsDirty(errorsRef.current);
    },
    validateFields,
    validate$,
    _setFieldTouched,
    _setFieldTouchedState,
    _registerField,
    _unregisterField,
  };

  return formCtx;
}
