// @byted-service/timer/lib/helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatOptions = exports.formatTimeoutArgs = exports.convertTime = exports.convertHRTime = exports.time = void 0;
const consts_1 = require("./consts");
const { s, second, ms, millisecond, μs, microsecond, ns, nanosecond } = consts_1.Units;
const time = (startTime) => {
    const cache = {};
    return (unit) => cache[unit] || (cache[unit] = (0, exports.convertHRTime)(startTime, unit));
};
exports.time = time;
const convertHRTime = (time, unit) => {
    switch (unit) {
        case s:
        case second:
            return time[0];
        case ms:
        case millisecond:
            return (time[0] * 1e3 + time[1] / 1e6) >> 0;
        case μs:
        case microsecond:
            return (time[0] * 1e6 + time[1] / 1e3) >> 0;
        case ns:
        case nanosecond:
            return time[0] * 1e9 + time[1];
        default:
            throw new Error(`Unknown unit: ${unit}`);
    }
};
exports.convertHRTime = convertHRTime;
const units = [
    [s, 1e9],
    [second, 1e9],
    [ms, 1e6],
    [millisecond, 1e6],
    [μs, 1e3],
    [microsecond, 1e3],
    [ns, 1],
    [nanosecond, 1],
];
const convertUnitMap = new Map();
for (const [from, fromWeight] of units) {
    for (const [to, toWeight] of units) {
        convertUnitMap.set(`${from}-${to}`, fromWeight / toWeight);
    }
}
const convertTime = (time, from, to) => {
    if (typeof time !== 'number') {
        return null;
    }
    if (from === to) {
        return time;
    }
    const convertUnit = convertUnitMap.get(`${from}-${to}`);
    if (typeof convertUnit !== 'number') {
        throw new Error(`Can't convert time, time: ${time}, from: ${from}, to: ${to}`);
    }
    return Math.round(time * convertUnit);
};
exports.convertTime = convertTime;
const formatTimeoutArgs = (unit, timeout, defaultUnit, defaultTimeout) => {
    if (unit && !consts_1.Units[unit]) {
        throw new Error(`Unknown unit: ${unit}`);
    }
    return {
        unit: unit || defaultUnit,
        timeout: typeof timeout === 'number'
            ? timeout
            : unit
                ? (0, exports.convertTime)(defaultTimeout, defaultUnit, unit)
                : defaultTimeout,
    };
};
exports.formatTimeoutArgs = formatTimeoutArgs;
const formatOptions = (options) => {
    if (typeof options === 'number') {
        return { unit: ms, timeout: options };
    }
    else if (options) {
        if (options.unit && !consts_1.Units[options.unit]) {
            throw new Error(`Unknown unit: ${options.unit}`);
        }
        return options;
    }
    else {
        return { unit: ms, timeout: null };
    }
};
exports.formatOptions = formatOptions;
