// @byted-service/timer/lib/timer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Timer = void 0;
const helper_1 = require("./helper");
const consts_1 = require("./consts");
const { ms, millisecond } = consts_1.Units;
class Timer {
    constructor(options) {
        this.startTime = null;
        this.stopTime = null;
        this.nowTime = ((unit) => (0, helper_1.convertHRTime)(process.hrtime(), unit));
        this.startTimestamp = 0;
        this.stopTimestamp = 0;
        this.timelineMap = {};
        this._timelines = [];
        const { unit, timeout, startOnCreate } = (0, helper_1.formatOptions)(options);
        this.unit = unit || ms;
        this.timeout = typeof timeout === 'number' ? timeout : null;
        if (startOnCreate !== false) {
            this.start();
        }
    }
    start() {
        if (this.stopTime) {
            throw new Error('Timer already stoped');
        }
        if (!this.startTime) {
            this.startTime = (0, helper_1.time)(process.hrtime());
            this.startTimestamp = Date.now();
        }
        return this;
    }
    stop(unit) {
        if (!this.startTime) {
            throw new Error('Timer has not yet started');
        }
        if (!this.stopTime) {
            this.stopTime = (0, helper_1.time)(process.hrtime());
            this.stopTimestamp = Date.now();
        }
        const realUnit = unit || this.unit;
        return realUnit === ms || realUnit === millisecond
            ? this.stopTimestamp - this.startTimestamp
            : this.stopTime(realUnit) - this.startTime(realUnit);
    }
    cost(unit) {
        if (!this.startTime) {
            throw new Error('Timer has not yet started');
        }
        const realUnit = unit || this.unit;
        const cost = realUnit === ms || realUnit === millisecond
            ? this.stopTime
                ? this.stopTimestamp - this.startTimestamp
                : Date.now() - this.startTimestamp
            : this.stopTime
                ? this.stopTime(realUnit) - this.startTime(realUnit)
                : this.nowTime(realUnit) - this.startTime(realUnit);
        return cost;
    }
    remain(unit, timeout) {
        if (!this.startTime) {
            throw new Error('Timer has not yet started');
        }
        if (this.stopTime) {
            return 0;
        }
        const { unit: realUnit, timeout: realTimeout } = (0, helper_1.formatTimeoutArgs)(unit, timeout, this.unit, this.timeout);
        if (typeof realTimeout !== 'number') {
            return Infinity;
        }
        const remain = realTimeout -
            (realUnit === ms || realUnit === millisecond
                ? Date.now() - this.startTimestamp
                : this.nowTime(realUnit) - this.startTime(realUnit));
        return remain < 0 ? 0 : remain;
    }
    async call(fn, options) {
        if (!this.startTime) {
            throw new Error('Timer has not yet started');
        }
        if (this.stopTime) {
            throw new Error('Timer already stoped');
        }
        const { times, fnTimeout, unit, finalError } = options || {};
        const realTimes = typeof times === 'number' ? times : 1;
        if (realTimes < 1) {
            throw new Error(`options.times must >= 1. Received: ${realTimes}`);
        }
        const timeoutArgs = (0, helper_1.formatTimeoutArgs)(unit, fnTimeout, this.unit, null);
        const realUnit = timeoutArgs.unit;
        let realFnTimeout = timeoutArgs.timeout;
        if (typeof realFnTimeout === 'number' && realFnTimeout <= 0) {
            throw new Error(`options.fnTimeout must > 0. Received: ${realFnTimeout}`);
        }
        let remain = this.remain(realUnit);
        if (remain <= 0) {
            throw new Error('Timer already stoped');
        }
        if (typeof realFnTimeout !== 'number') {
            realFnTimeout = remain;
        }
        let i = 0;
        let error = null;
        while (i < realTimes && remain > 0) {
            let timeout = (realFnTimeout > remain ? remain : realFnTimeout);
            timeout = Number.isFinite(timeout) ? timeout : undefined;
            let p = null;
            try {
                p = fn(timeout, i);
                return await p;
            }
            catch (err) {
                // 如果 p 都没有产生，说明是同步错误，对同步错误进行重试是没有意义的，直接 throw
                if (!p) {
                    throw err;
                }
                remain = this.remain(realUnit);
                i++;
                error = err;
            }
        }
        if (finalError) {
            if (typeof finalError === 'function') {
                throw finalError(error);
            }
            else {
                throw finalError;
            }
        }
        throw error;
    }
    child(options) {
        if (!this.startTime) {
            throw new Error('Timer has not yet started');
        }
        if (this.stopTime) {
            throw new Error('Timer already stoped');
        }
        const formattedOptions = (0, helper_1.formatOptions)(options);
        const unit = formattedOptions.unit || this.unit;
        let timeout = formattedOptions.timeout;
        const remain = this.remain(unit);
        if (typeof timeout === 'number') {
            timeout = timeout > remain ? remain : timeout;
        }
        else {
            timeout = remain;
        }
        if (!Number.isFinite(timeout)) {
            timeout = null;
        }
        formattedOptions.unit = unit;
        formattedOptions.timeout = timeout;
        return new Timer(formattedOptions);
    }
    hasTimeline(key) {
        return Boolean(this.timelineMap[key]);
    }
    delTimeline(key) {
        const timeline = this.timelineMap[key];
        if (timeline) {
            delete this.timelineMap[key];
            this._timelines = this._timelines.filter((t) => t !== timeline);
        }
        return timeline;
    }
    timeline(key, unit) {
        const realUnit = unit || this.unit;
        if (!this.startTime) {
            throw new Error('Timer has not yet started');
        }
        if (this.stopTime) {
            throw new Error('Timer already stoped');
        }
        const timeline = this.timelineMap[key];
        if (timeline) {
            timeline.stopTime = timeline.stopTime || (0, helper_1.time)(process.hrtime());
            return timeline.stopTime(realUnit) - timeline.startTime(realUnit);
        }
        else {
            const timeline = {
                key,
                startTime: (0, helper_1.time)(process.hrtime()),
                stopTime: null,
            };
            this.timelineMap[key] = timeline;
            this._timelines.push(timeline);
            return 0;
        }
    }
    timelines(unit) {
        const realUnit = unit || this.unit;
        return this._timelines.map(({ key, startTime, stopTime }) => {
            const start = startTime(realUnit);
            const stop = (stopTime || this.nowTime)(realUnit);
            return {
                key,
                cost: stop - start,
                start,
                stop,
                unit: realUnit,
                stoped: Boolean(stopTime),
            };
        });
    }
    /** @deprecated */
    unref() {
        return this;
    }
}
exports.Timer = Timer;
