// @universe-design/react/lib/menu/internal-menu/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Divider", {
  enumerable: true,
  get: function () {
    return _Divider.default;
  }
});
Object.defineProperty(exports, "MenuContext", {
  enumerable: true,
  get: function () {
    return _MenuContext.RcMenuContext;
  }
});
Object.defineProperty(exports, "RcMenuItem", {
  enumerable: true,
  get: function () {
    return _MenuItem.default;
  }
});
Object.defineProperty(exports, "RcMenuItemGroup", {
  enumerable: true,
  get: function () {
    return _MenuItemGroup.default;
  }
});
Object.defineProperty(exports, "RcSubMenu", {
  enumerable: true,
  get: function () {
    return _SubMenu.default;
  }
});
exports.useFullPath = exports.default = void 0;
var _Menu = _interopRequireDefault(require("./Menu"));
var _MenuItem = _interopRequireDefault(require("./MenuItem"));
var _SubMenu = _interopRequireDefault(require("./SubMenu"));
var _MenuItemGroup = _interopRequireDefault(require("./MenuItemGroup"));
var _PathContext = require("./context/PathContext");
var _Divider = _interopRequireDefault(require("./Divider"));
var _MenuContext = require("./context/MenuContext");
/** @private Only used for antd internal. Do not use in your production. */
const useFullPath = exports.useFullPath = _PathContext.useFullPath;
const ExportMenu = _Menu.default;
ExportMenu.Item = _MenuItem.default;
ExportMenu.SubMenu = _SubMenu.default;
ExportMenu.ItemGroup = _MenuItemGroup.default;
ExportMenu.Divider = _Divider.default;
var _default = exports.default = ExportMenu;
