// @universe-design/react/lib/menu/internal-menu/MenuItem.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _rcOverflow = _interopRequireDefault(require("rc-overflow"));
var _warning = _interopRequireDefault(require("@universe-design/rc-util/lib/warning"));
var _KeyCode = _interopRequireDefault(require("@universe-design/rc-util/lib/KeyCode"));
var _omit = _interopRequireDefault(require("@universe-design/rc-util/lib/omit"));
var _MenuContext = require("./context/MenuContext");
var _useActive = _interopRequireDefault(require("./hooks/useActive"));
var _warnUtil = require("./utils/warnUtil");
var _Icon = _interopRequireDefault(require("./Icon"));
var _useDirectionStyle = _interopRequireDefault(require("./hooks/useDirectionStyle"));
var _PathContext = require("./context/PathContext");
var _IdContext = require("./context/IdContext");
// Since Menu event provide the `info.item` which point to the MenuItem node instance.
// We have to use class component here.
// This should be removed from doc & api in future.
class LegacyMenuItem extends React.Component {
  render() {
    const {
      title,
      attribute,
      elementRef,
      ...restProps
    } = this.props;
    const passedProps = (0, _omit.default)(restProps, ['eventKey']);
    process.env.NODE_ENV !== "production" ? (0, _warning.default)(!attribute, '`attribute` of Menu.Item is deprecated. Please pass attribute directly.') : void 0;
    return /*#__PURE__*/React.createElement(_rcOverflow.default.Item, (0, _extends2.default)({}, attribute, {
      title: typeof title === 'string' ? title : undefined
    }, passedProps, {
      ref: elementRef
    }));
  }
}

/**
 * Real Menu Item component
 */
const InternalMenuItem = props => {
  const {
    style,
    className,
    eventKey,
    warnKey,
    disabled,
    itemIcon,
    children,
    // Aria
    role,
    // Active
    onMouseEnter,
    onMouseLeave,
    onClick,
    onKeyDown,
    onFocus,
    ...restProps
  } = props;
  const domDataId = (0, _IdContext.useMenuId)(eventKey);
  const {
    prefixCls,
    onItemClick,
    disabled: contextDisabled,
    overflowDisabled,
    // Icon
    itemIcon: contextItemIcon,
    // Select
    selectedKeys,
    // Active
    onActive
  } = React.useContext(_MenuContext.RcMenuContext);
  const itemCls = `${prefixCls}-item`;
  const legacyMenuItemRef = React.useRef();
  const elementRef = React.useRef();
  const mergedDisabled = Boolean(contextDisabled || disabled);
  const connectedKeys = (0, _PathContext.useFullPath)(eventKey);

  // ================================ Warn ================================
  if (process.env.NODE_ENV !== 'production' && warnKey) {
    process.env.NODE_ENV !== "production" ? (0, _warning.default)(false, 'MenuItem should not leave undefined `key`.') : void 0;
  }

  // ============================= Info =============================
  const getEventInfo = e => ({
    key: eventKey,
    // Note: For legacy code is reversed which not like other antd component
    keyPath: [...connectedKeys].reverse(),
    item: legacyMenuItemRef.current,
    domEvent: e
  });

  // ============================= Icon =============================

  // ============================ Active ============================
  const {
    active,
    ...activeProps
  } = (0, _useActive.default)(eventKey, mergedDisabled, onMouseEnter, onMouseLeave);

  // ============================ Select ============================
  const selected = selectedKeys.includes(eventKey);

  // ======================== DirectionStyle ========================
  const directionStyle = (0, _useDirectionStyle.default)(connectedKeys.length);

  // ============================ Events ============================

  /**
   * Used for accessibility. Helper will focus element without key board.
   * We should manually trigger an active
   */

  // ============================ Render ============================
  const optionRoleProps = {};
  if (props.role === 'option') {
    optionRoleProps['aria-selected'] = selected;
  }
  return /*#__PURE__*/React.createElement(LegacyMenuItem, (0, _extends2.default)({
    ref: legacyMenuItemRef,
    elementRef: elementRef,
    role: role === null ? 'none' : role || 'menuitem',
    tabIndex: disabled ? null : -1,
    "data-menu-id": overflowDisabled && domDataId ? null : domDataId
  }, restProps, activeProps, optionRoleProps, {
    component: "li",
    "aria-disabled": disabled,
    style: {
      ...directionStyle,
      ...style
    },
    className: (0, _clsx.default)(itemCls, {
      [`${itemCls}-active`]: active,
      [`${itemCls}-selected`]: selected,
      [`${itemCls}-disabled`]: mergedDisabled
    }, className),
    onClick: e => {
      if (mergedDisabled) {
        return;
      }
      const info = getEventInfo(e);
      onClick === null || onClick === void 0 ? void 0 : onClick((0, _warnUtil.warnItemProp)(info));
      onItemClick(info);
    },
    onKeyDown: e => {
      onKeyDown === null || onKeyDown === void 0 ? void 0 : onKeyDown(e);
      if (e.which === _KeyCode.default.ENTER) {
        const info = getEventInfo(e);

        // Legacy. Key will also trigger click event
        onClick === null || onClick === void 0 ? void 0 : onClick((0, _warnUtil.warnItemProp)(info));
        onItemClick(info);
      }
    },
    onFocus: e => {
      onActive(eventKey);
      onFocus === null || onFocus === void 0 ? void 0 : onFocus(e);
    }
  }), children, /*#__PURE__*/React.createElement(_Icon.default, {
    props: {
      ...props,
      isSelected: selected
    },
    icon: itemIcon || contextItemIcon
  }));
};
function RcMenuItem(props) {
  const {
    eventKey
  } = props;

  // ==================== Record KeyPath ====================
  const measure = (0, _PathContext.useMeasure)();
  const connectedKeyPath = (0, _PathContext.useFullPath)(eventKey);

  // eslint-disable-next-line consistent-return
  React.useEffect(() => {
    if (measure) {
      measure.registerPath(eventKey, connectedKeyPath);
      return () => {
        measure.unregisterPath(eventKey, connectedKeyPath);
      };
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [connectedKeyPath]);
  if (measure) {
    return null;
  }

  // ======================== Render ========================
  return /*#__PURE__*/React.createElement(InternalMenuItem, props);
}
var _default = exports.default = RcMenuItem;
