// @universe-design/react/es/shared/components/trigger/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React, { forwardRef, useCallback, useImperativeHandle, useMemo, useRef, useState } from 'react';
import RcTrigger from '@universe-design/rc-trigger';
import merge from 'lodash/merge';
import cloneDeep from 'lodash/cloneDeep';
import pick from 'lodash/pick';
import { useControlledState } from '@universe-design/react-utilities';
import { useClassPrefix } from '../../../config-provider/hooks';
import { isFunction, __SCREENSHOT__, devWarning, warning } from '../../../util';
import { useIsomorphicLayoutEffect as useLayoutEffect } from '../../../hooks';
import { useTriggerDisabled } from './modifiers/useTriggerDisabled';
import clsx from 'clsx';
import Popup from './Popup';
import { useTriggerObserver } from './modifiers/useTriggerObserver';
import composeRefs from '../../../util/composeRefs';
import { isOverflowed as defaultIsOverflow } from '../../../util/dom';
import { usePopupPortalContainer } from '@universe-design/react-provider';

/**
 * @docgen
 * @public
 */

export function getAlignFromPlacement(builtinPlacements, placementStr, align) {
  const baseAlign = builtinPlacements[placementStr] || {};
  return merge(cloneDeep(baseAlign), align);
}
export const defaults = {
  mouseEnterDelay: 100,
  mouseLeaveDelay: 100,
  destroyOnHide: {
    keepParent: false
  },
  popupAlign: {}
};
export const UDTrigger = /*#__PURE__*/forwardRef((props, ref) => {
  const rootPrefixCls = useClassPrefix();
  const {
    prefixCls = `${rootPrefixCls}-trigger`,
    disabled = false,
    popupMotion = 'slide-up',
    arrow: arrowProps,
    mouseEnterDelay = defaults.mouseEnterDelay,
    mouseLeaveDelay = defaults.mouseLeaveDelay,
    popupPlacement,
    popupAlign: popupAlignProp = defaults.popupAlign,
    builtinPlacements,
    modifiers = {},
    destroyOnHide = defaults.destroyOnHide,
    visible: visibleProp,
    checkOverflow = false,
    defaultVisible = false,
    onVisibleChange,
    openClassName = `${prefixCls}-open`,
    className,
    udTriggerRef,
    getPopupContainer: getPopupContainerProp,
    popupClassName,
    children,
    __screenshotAlwaysOpenEnabled = false,
    popupStopPropagation,
    ...restProps
  } = props;
  const {
    popupOverflowAuto: enablePopupOverflowAuto = false,
    forcePopupAlign: enableforcePopupAlign = false
  } = modifiers;

  // getPopupContainer
  const ctxGetPopupContainer = usePopupPortalContainer();

  // 获取用于对齐的 popupAlign 配置，逻辑从 rc-trigger 提升
  const popupAlign = useMemo(() => {
    if (popupPlacement && builtinPlacements) {
      return getAlignFromPlacement(builtinPlacements, popupPlacement, popupAlignProp);
    }
    return popupAlignProp;
  }, [builtinPlacements, popupAlignProp, popupPlacement]);

  // visible
  const [visible, setVisible] = useControlledState(defaultVisible, visibleProp, {
    onChange: onVisibleChange
  });
  const triggerRef = useRef(null);
  const checkVisible = useCallback(_visible => {
    if (_visible && checkOverflow) {
      var _triggerRef$current;
      const child = (_triggerRef$current = triggerRef.current) === null || _triggerRef$current === void 0 ? void 0 : _triggerRef$current.getRootDomNode();
      const overflowChecker = typeof checkOverflow === 'function' ? checkOverflow : defaultIsOverflow;
      if (!child || !overflowChecker(child)) {
        setVisible(false);
        return;
      }
    }
    setVisible(_visible);
  }, [checkOverflow, setVisible]);
  __SCREENSHOT__ && __screenshotAlwaysOpenEnabled &&
  // eslint-disable-next-line react-hooks/rules-of-hooks
  React.useEffect(() => {
    var _triggerRef$current2;
    (_triggerRef$current2 = triggerRef.current) === null || _triggerRef$current2 === void 0 ? void 0 : _triggerRef$current2.forcePopupAlign();
    checkVisible(true);
  });

  // destroy
  let destroyPopupOnHide = false;
  let autoDestroy = false;
  if (typeof destroyOnHide === 'boolean') {
    destroyPopupOnHide = destroyOnHide;
  } else if (destroyOnHide && typeof destroyOnHide === 'object') {
    const {
      keepParent
    } = destroyOnHide;
    destroyPopupOnHide = keepParent === true;
    autoDestroy = keepParent === false;
  }

  // 预处理 props
  const rcTriggerProps = {
    ...restProps,
    children: isFunction(children) ? children({
      visible
    }) : children,
    popupClassName: clsx(popupClassName, mouseLeaveDelay === 0 && `${prefixCls}-no-entry`),
    // forceRender 的优先级大于 destroyPopupOnHide
    // 俩都传了 true 的时候, 相当于 hide 后不会销毁
    destroyPopupOnHide,
    autoDestroy,
    mouseEnterDelay: mouseEnterDelay / 1000,
    mouseLeaveDelay: mouseLeaveDelay / 1000,
    prefixCls,
    popupMotion: {
      motionName: `${rootPrefixCls}-${popupMotion}`
    },
    popupAlign,
    builtinPlacements,
    popupPlacement,
    popupVisible: visible,
    /**
     * 这里现在有点问题，rc-trigger onPopupVisibleChange 在 visible false 的时候不能触发，所以 checkOverflow 开启时，如果子元素从 overflow 变化到不 overflow，鼠标移出 trigger 区域，子元素再从不 overflow 变成 overflow，tooltip 会再次出现
     */
    onPopupVisibleChange: checkVisible,
    getPopupContainer: getPopupContainerProp !== null && getPopupContainerProp !== void 0 ? getPopupContainerProp : ctxGetPopupContainer
  };

  // 禁用 trigger
  const [newProps] = useTriggerDisabled(rcTriggerProps, {
    disabled
  });
  useImperativeHandle(udTriggerRef, () => ({
    visible,
    setVisible,
    forcePopupAlign: () => {
      var _triggerRef$current3;
      triggerRef === null || triggerRef === void 0 ? void 0 : (_triggerRef$current3 = triggerRef.current) === null || _triggerRef$current3 === void 0 ? void 0 : _triggerRef$current3.forcePopupAlign();
    }
  }), [visible, setVisible]);
  const [readyObserver, setReadyObserver] = useState(false);
  const startObserver = useCallback(() => setReadyObserver(true), []);
  const [childSize, popupSize, viewPointRect] = useTriggerObserver({
    triggerRef,
    popupAlign,
    readyObserver,
    enableforcePopupAlign,
    enablePopupOverflowAuto
  });
  useLayoutEffect(() => {
    if (!visible && readyObserver) {
      setReadyObserver(false);
    }
  }, [visible, readyObserver]);
  const composeTrigger = useMemo(() => composeRefs(triggerRef, ref), [triggerRef, ref]);
  if (disabled) {
    return /*#__PURE__*/React.createElement(React.Fragment, null, rcTriggerProps.children);
  }

  // TODO 临时逻辑，处理 menu eventKey 透传
  process.env.NODE_ENV !== "production" ? warning(!('eventKey' in newProps), devWarning('Trigger',
  // eslint-disable-next-line max-len
  'The `eventKey` property is detected, maybe you wrap the popper elements such as <Tooltip /> or <Popover /> outside the <MenuItem /> or <SubMenu />, it is recommended to move it to the inside')) : void 0;
  newProps.children = /*#__PURE__*/React.cloneElement(newProps.children, pick(newProps, ['eventKey']));
  return /*#__PURE__*/React.createElement(RcTrigger, _extends({}, newProps, {
    popup: /*#__PURE__*/React.createElement(Popup, {
      enablePopupOverflowAuto: enablePopupOverflowAuto,
      childSize: childSize,
      popupSize: popupSize,
      viewPointRect: viewPointRect,
      prefixCls: newProps.prefixCls,
      popup: newProps.popup,
      triggerRef: triggerRef,
      popupAlign: popupAlign,
      arrow: arrowProps,
      popupStopPropagation: popupStopPropagation
    }),
    className: clsx(className, visible && openClassName),
    popupVisible: visible,
    onPopupAlign: startObserver,
    ref: composeTrigger
  }));
});
if (process.env.NODE_ENV !== "production") {
  UDTrigger.displayName = 'UDTrigger';
}
export * from './types';
