// @universe-design/react/es/tag/Tag.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.u10pi13n{position:relative;}
import u5jxv71 from "./../.stylex/u5jxv71";
//.u2gz7yw::after{display:block;}
import u1lumzkm from "./../.stylex/u1lumzkm";
//.u1mdlcz9::after{position:absolute;}
import u11yfbjh from "./../.stylex/u11yfbjh";
//.u7na3ze::after{width:24px;}
import u8o0f2y from "./../.stylex/u8o0f2y";
//.u1shord1::after{height:24px;}
import ugk9fm9 from "./../.stylex/ugk9fm9";
//.u13zj6fq::after{content:"";}
import u7grb02 from "./../.stylex/u7grb02";
//.u9eqz9b::after{top:calc(50% - 12px);}
import u1g1yqg3 from "./../.stylex/u1g1yqg3";
//.ua2bdqt::after{left:-2px;}
import u3rhfi7 from "./../.stylex/u3rhfi7";
//.u1a3p1vp{overflow:hidden;}
import ux645fm from "./../.stylex/ux645fm";
//.u161knb0{padding-left:2px;}
import u1km0f1u from "./../.stylex/u1km0f1u";
//.u1n0apg3{font-size:8px;}
import uahyb5n from "./../.stylex/uahyb5n";
//.u1hzsgrp{margin-left:2px;}
import u1agd5ph from "./../.stylex/u1agd5ph";
//.u1hkdmbb:hover{opacity:0.5;}
import unf6cqd from "./../.stylex/unf6cqd";
//.uc5e8lb:pressed{opacity:0.5;}
import u1c0bjt1 from "./../.stylex/u1c0bjt1";
//.u1k6fduh{cursor:pointer;}
import u12o6wmt from "./../.stylex/u12o6wmt";
//.u1hfsmdc{opacity:0.6;}
import u15krktc from "./../.stylex/u15krktc";
//.u15pjodv:hover{opacity:1;}
import u1dhzg2w from "./../.stylex/u1dhzg2w";
//.u9zk3pm:pressed{opacity:1;}
import u6veevw from "./../.stylex/u6veevw";
//.ujf1xye{margin-left:4px;}
import uweqc5r from "./../.stylex/uweqc5r";
//.udayyex{opacity:0.5;}
import uiedlft from "./../.stylex/uiedlft";
//.ui64zpg{flex-shrink:0;}
import uidkx17 from "./../.stylex/uidkx17";
//.u1q9cn9r{padding-right:10.666666666666666px;}
import ukrpgfj from "./../.stylex/ukrpgfj";
//.u101vrxs{padding-right:6.666666666666667px;}
import u11l27z2 from "./../.stylex/u11l27z2";
//.u1a5o2fe{padding-right:5.333333333333333px;}
import u1ei76u2 from "./../.stylex/u1ei76u2";
//.u1szoe96{width:32px;}
import u1y9lr4k from "./../.stylex/u1y9lr4k";
//.uq4mcun{width:24px;}
import uymufub from "./../.stylex/uymufub";
//.u64fuq3{width:20px;}
import uevmt6d from "./../.stylex/uevmt6d";
//.ujw5fx7{width:16px;}
import ut0j16j from "./../.stylex/ut0j16j";
//.u1nbblvp{padding-top:0px;}
import u165peiw from "./../.stylex/u165peiw";
//.uifp7yv{padding-right:0px;}
import u24on9z from "./../.stylex/u24on9z";
//.u1ov4xf1{padding-bottom:0px;}
import uu1oa8u from "./../.stylex/uu1oa8u";
//.u1asdtw4{padding-left:0px;}
import u1pwi6nc from "./../.stylex/u1pwi6nc";
//.u4d9j23{justify-content:center;}
import ut7q8es from "./../.stylex/ut7q8es";
//.u4ybsrx{font-size:16px;}
import u1o367ev from "./../.stylex/u1o367ev";
//.u1d2rq10{height:32px;}
import ujixoer from "./../.stylex/ujixoer";
//.u19lj068{padding-right:8px;}
import u441rjq from "./../.stylex/u441rjq";
//.u177v4lu{padding-left:8px;}
import um3isuf from "./../.stylex/um3isuf";
//.uses1vf{font-size:14px;}
import u19bueuo from "./../.stylex/u19bueuo";
//.urvgh55{height:24px;}
import ud7edr6 from "./../.stylex/ud7edr6";
//.u1ugzwwg{font-size:12px;}
import u283mfx from "./../.stylex/u283mfx";
//.ujamq6b{height:20px;}
import u1ny05fg from "./../.stylex/u1ny05fg";
//.u1sgzk6v{padding-right:6px;}
import u1hkv3ak from "./../.stylex/u1hkv3ak";
//.u1bg5zqg{padding-left:6px;}
import u1m9gb09 from "./../.stylex/u1m9gb09";
//.unmn6fi{font-size:10px;}
import u16uuw9n from "./../.stylex/u16uuw9n";
//.ud461yt{height:16px;}
import ua0xsel from "./../.stylex/ua0xsel";
//.u1g0x7ka{padding-top:0;}
import u1fdqv2r from "./../.stylex/u1fdqv2r";
//.u8wuabp{padding-right:4px;}
import upmcqac from "./../.stylex/upmcqac";
//.u1qch9an{padding-bottom:0;}
import u6qvdto from "./../.stylex/u6qvdto";
//.uycuoez{padding-left:4px;}
import u1ram9zc from "./../.stylex/u1ram9zc";
//.udrzuqr{cursor:not-allowed;}
import u1vjduc2 from "./../.stylex/u1vjduc2";
//.u113hnb5{cursor:text;}
import u1qr18xr from "./../.stylex/u1qr18xr";
//.utuwxu6{display:inline-flex;}
import ug0oh3b from "./../.stylex/ug0oh3b";
//.u122n59{align-items:center;}
import ubyhozq from "./../.stylex/ubyhozq";
//.uz5stix{white-space:nowrap;}
import uugkpzq from "./../.stylex/uugkpzq";
//.u7o03ib{transition-property:all;}
import u18eu6n2 from "./../.stylex/u18eu6n2";
//.u15l5c9n{transition-duration:0.2s;}
import uf7fnhg from "./../.stylex/uf7fnhg";
//.u6dzj5z{max-width:100%;}
import u9hwb1v from "./../.stylex/u9hwb1v";
import CloseBoldOutlined from '@universe-design/icons-react/CloseBoldOutlined';
import clsx from 'clsx';
import React, { cloneElement, useRef } from 'react';
import stylex from '@universe-design/stylex';
import { useControlledState } from '@universe-design/react-utilities';
import { cssBlockBase, cssComponentBase, cssTypographyBody0, cssTypographyBody2, cssTypographyCaption0, cssTypographyCaption1Hack, cssTypographyHeadline, cssTypographyTitle4, cssTypographyTitle5, sxBorderRadius, sxOverflowEllipsis, sxTransitionTimingFunction } from '../stylex';
import { useCssScope } from '../config-provider/hooks';
import { devWarning, getSizeAbbr, isNonNullable, warning } from '../util';
import { isAvatar } from '../avatar/common';
import { useTextOverflowPopperSidecar } from '../config-provider/hooks/useTextOverflowPopperSidecar';
import { isHorizontalOverflowed } from '../util/dom';
import './style';

// xs
const tagSizeConfigXsFontSize = 10;
const tagSizeConfigXsCloseIconSize = 8;
const tagSizeConfigXsHeight = 16;
const tagSizeConfigXsPaddingX = 4;

// sm
const tagSizeConfigSmFontSize = 12;
const tagSizeConfigSmCloseIconSize = 10;
const tagSizeConfigSmHeight = 20;
const tagSizeConfigSmPaddingX = 6;

// md
const tagSizeConfigMdFontSize = 14;
const tagSizeConfigMdCloseIconSize = 12;
const tagSizeConfigMdHeight = 24;
const tagSizeConfigMdPaddingX = 6;

// lg
const tagSizeConfigLgFontSize = 16;
const tagSizeConfigLgCloseIconSize = 14;
const tagSizeConfigLgHeight = 32;
const tagSizeConfigLgPaddingX = 8;
const sxTag = {
  root: {
    base: stylex.compose(cssComponentBase, sxBorderRadius.xs, {
      "props": {
        "Bceei9c": u1qr18xr,
        "mc9l5x": ug0oh3b,
        "Bt984gj": ubyhozq,
        "Huce71": uugkpzq,
        "Bmy1vo4": u18eu6n2,
        "B3o57yi": uf7fnhg,
        "B2u0y6b": u9hwb1v
      }
    }, sxTransitionTimingFunction.standard),
    disabled: {
      "props": {
        "Bceei9c": u1vjduc2
      }
    },
    xs: {
      "props": {
        "Be2twd7": u16uuw9n,
        "Bqenvij": ua0xsel,
        "z8tnut": u1fdqv2r,
        "z189sj": upmcqac,
        "Byoj8tv": u6qvdto,
        "uwmqm3": u1ram9zc
      }
    },
    sm: {
      "props": {
        "Be2twd7": u283mfx,
        "Bqenvij": u1ny05fg,
        "z8tnut": u1fdqv2r,
        "z189sj": u1hkv3ak,
        "Byoj8tv": u6qvdto,
        "uwmqm3": u1m9gb09
      }
    },
    md: {
      "props": {
        "Be2twd7": u19bueuo,
        "Bqenvij": ud7edr6,
        "z8tnut": u1fdqv2r,
        "z189sj": u1hkv3ak,
        "Byoj8tv": u6qvdto,
        "uwmqm3": u1m9gb09
      }
    },
    lg: {
      "props": {
        "Be2twd7": u1o367ev,
        "Bqenvij": ujixoer,
        "z8tnut": u1fdqv2r,
        "z189sj": u441rjq,
        "Byoj8tv": u6qvdto,
        "uwmqm3": um3isuf
      }
    },
    square: {
      base: {
        "props": {
          "z8tnut": u165peiw,
          "z189sj": u24on9z,
          "Byoj8tv": uu1oa8u,
          "uwmqm3": u1pwi6nc,
          "Brf1p80": ut7q8es
        }
      },
      xs: {
        "props": {
          "a9b677": ut0j16j
        }
      },
      sm: {
        "props": {
          "a9b677": uevmt6d
        }
      },
      md: {
        "props": {
          "a9b677": uymufub
        }
      },
      lg: {
        "props": {
          "a9b677": u1y9lr4k
        }
      }
    },
    round: {
      base: stylex.compose({
        "props": {
          "z8tnut": u1fdqv2r,
          "z189sj": u441rjq,
          "Byoj8tv": u6qvdto,
          "uwmqm3": um3isuf
        }
      }, sxBorderRadius.xl),
      xs: {
        "props": {
          "z189sj": u1ei76u2
        }
      },
      sm: {
        "props": {
          "z189sj": u11l27z2
        }
      },
      md: {
        "props": {
          "z189sj": u441rjq
        }
      },
      lg: {
        "props": {
          "z189sj": ukrpgfj
        }
      }
    }
  },
  statusIcon: {
    base: {
      "props": {
        "Bnnss6s": uidkx17
      }
    },
    xs: {
      "props": {
        "Be2twd7": u283mfx
      }
    }
  },
  content: {
    base: stylex.compose(cssBlockBase, sxOverflowEllipsis),
    disabled: {
      "props": {
        "abs64n": uiedlft
      }
    },
    xs: cssTypographyCaption1Hack,
    sm: cssTypographyBody2,
    md: cssTypographyBody0,
    lg: cssTypographyTitle5,
    status: {
      xs: cssTypographyCaption1Hack,
      sm: cssTypographyCaption0,
      md: cssTypographyHeadline,
      lg: cssTypographyTitle4
    },
    hasIcon: {
      "props": {
        "Frg6f3": uweqc5r
      }
    }
  },
  closeIcon: {
    base: stylex.compose({
      "props": {
        "Bceei9c": u12o6wmt,
        "Frg6f3": uweqc5r,
        "abs64n": u15krktc,
        "Bmy1vo4": u18eu6n2,
        "B3o57yi": uf7fnhg,
        "B7zu5sd": u1dhzg2w,
        "Bthtq7b": u6veevw
      }
    }, sxTransitionTimingFunction.standard),
    disabled: {
      "props": {
        "abs64n": uiedlft,
        "B7zu5sd": unf6cqd,
        "Bthtq7b": u1c0bjt1
      }
    },
    xs: {
      "props": {
        "Be2twd7": uahyb5n,
        "Frg6f3": u1agd5ph
      }
    },
    sm: {
      "props": {
        "Be2twd7": u16uuw9n
      }
    },
    md: {
      "props": {
        "Be2twd7": u283mfx
      }
    },
    lg: {
      "props": {
        "Be2twd7": u19bueuo
      }
    }
  }
};
function useTagCssScope() {
  return useCssScope('tag');
}

/**
 * 废弃的 color 命名，以及新命名
 */
const CompatibleColorMap = {
  'neutral-select': 'neutral-option',
  'neutral-light': 'neutral-solid'
};
const TagPresetColorStatusEnum = [
// Status Tag
'waiting', 'success', 'info', 'warning', 'error', 'missing'];
export const TagPresetColorEnum = [
// Neutral
'neutral', 'neutral-solid', 'neutral-option', 'neutral-ghost', 'neutral-ghost-transparent',
// Attribute Tag
'blue', 'wathet', 'turquoise', 'green', 'lime', 'yellow', 'orange', 'red', 'violet', 'purple', 'indigo', 'carmine',
// Solid Attribute Tag
'blue-solid', 'wathet-solid', 'turquoise-solid', 'green-solid', 'lime-solid', 'yellow-solid', 'orange-solid', 'red-solid', 'violet-solid', 'purple-solid', 'indigo-solid', 'carmine-solid',
// Option Tag
'blue-option', 'wathet-option', 'turquoise-option', 'green-option', 'lime-option', 'yellow-option', 'orange-option', 'red-option', 'violet-option', 'purple-option', 'indigo-option', 'carmine-option',
// Ghost Tag
'blue-ghost', 'carmine-ghost', 'red-ghost', ...TagPresetColorStatusEnum,
// Avatar Tag
'neutral-avatar', 'blue-avatar', 'orange-avatar'];

/**
 * @public
 */

const getPresetColor = color => {
  const compatibleColor = CompatibleColorMap[color];
  process.env.NODE_ENV !== "production" ? warning(Boolean(!compatibleColor), devWarning('Tag', `color value \`${color}\` is deprecated, please use \`${compatibleColor}\` instead.`)) : void 0;
  const _color = compatibleColor !== null && compatibleColor !== void 0 ? compatibleColor : color;
  if (TagPresetColorEnum.includes(_color)) {
    return [true, _color];
  }
  return [false, _color];
};

/**
 * @public
 */

const tagAvatarSize = {
  'extra-small': 12,
  small: 16,
  medium: 20,
  large: 28
};
function getAvatarSizeFromTagSize(size) {
  return tagAvatarSize[size];
}
export const Tag = /*#__PURE__*/React.forwardRef((props, ref) => {
  var _icon$props;
  const {
    color: colorProp = 'neutral',
    className,
    size: sizeProp = 'medium',
    icon,
    children,
    closable = false,
    visible: visibleProp,
    onClose,
    style,
    shape = 'rect',
    disabled = false,
    sx,
    ...restProps
  } = props;
  const size = getSizeAbbr(sizeProp);
  const cssScope = useTagCssScope();
  const TextOverflowPopper = useTextOverflowPopperSidecar();
  const [isPresetColor, color] = getPresetColor(colorProp);
  const tagStyle = {
    backgroundColor: color && !isPresetColor ? color : undefined,
    ...style
  };
  const [visible, setVisible] = useControlledState(true, visibleProp);
  const onCloseClick = React.useCallback(e => {
    // 这个逻辑参考其他开源框架，不过感觉有受控 visible 不需要这个 prevented 判断
    e.stopPropagation();
    onClose === null || onClose === void 0 ? void 0 : onClose(e);
    if (e.defaultPrevented) {
      return;
    }
    setVisible(false);
  }, [onClose, setVisible]);
  const contentCls = clsx(cssScope('__content'), icon && cssScope('__content-has-icon'));
  const contentSx = [sxTag.content.base, disabled && sxTag.content.disabled, icon && sxTag.content.hasIcon, icon && size === 'xs' && {
    "props": {
      "Frg6f3": u1agd5ph
    }
  }, size && (TagPresetColorStatusEnum.includes(color) ? sxTag.content.status[size] : sxTag.content[size])];
  const contentRef = useRef(null);
  /** extra-small 因 zoom 问题, 暂时关闭 Tooltip */
  const contentNode = (() => {
    // square tag 没有地方放 children 了
    if (!isNonNullable(children) || shape === 'square') {
      return null;
    }
    return /*#__PURE__*/React.createElement("div", {
      ref: contentRef,
      className: clsx(contentCls, stylex(contentSx))
    }, children);
  })();
  return visible ? /*#__PURE__*/React.createElement(TextOverflowPopper, {
    title: children,
    checkOverflow: () => {
      if (contentRef.current) {
        return isHorizontalOverflowed(contentRef.current);
      }
      return false;
    }
  }, /*#__PURE__*/React.createElement("span", _extends({
    ref: ref,
    className: clsx(cssScope(), className, size && cssScope(`-${size}`), shape && cssScope(`-${shape}`), cssScope(!isPresetColor ? '-other-color' : `-${color}`), icon && cssScope('-has-icon'), disabled && cssScope('-disabled'), stylex(sxTag.root.base, disabled && sxTag.root.disabled, size && sxTag.root[size], (shape === 'round' || shape === 'square') && sxTag.root[shape].base, size && (shape === 'round' || shape === 'square') && sxTag.root[shape][size], shape === 'round' && icon && {
      "props": {
        "z8tnut": u165peiw,
        "uwmqm3": u1km0f1u,
        "Byoj8tv": uu1oa8u
      }
    }, {
      "props": {
        "Bpg54ce": ux645fm
      }
    }, sx)),
    style: tagStyle
  }, restProps), icon ? /*#__PURE__*/cloneElement(icon, {
    className: clsx(cssScope('__status-icon'), icon === null || icon === void 0 ? void 0 : (_icon$props = icon.props) === null || _icon$props === void 0 ? void 0 : _icon$props.className, stylex(sxTag.statusIcon.base, size === 'xs' && sxTag.statusIcon[size])),
    ...(isAvatar(icon) ? {
      size: getAvatarSizeFromTagSize(sizeProp)
    } : {})
  }) : null, contentNode, closable && shape !== 'square' ? /*#__PURE__*/React.createElement(CloseBoldOutlined, {
    className: clsx(cssScope('__close-icon'), stylex(sxTag.closeIcon.base, disabled && sxTag.closeIcon.disabled, size && sxTag.closeIcon[size], {
      "props": {
        "qhf8xq": u5jxv71,
        "ap17g6": u1lumzkm,
        "E3zdtr": u11yfbjh,
        "a2br6o": u8o0f2y,
        "Dlnsje": ugk9fm9,
        "Bsft5z2": u7grb02,
        "bn5sak": u1g1yqg3,
        "Eqx8gd": u3rhfi7
      }
    })),
    onClick: onCloseClick
  }) : null)) : null;
});
if (process.env.NODE_ENV !== "production") {
  Tag.displayName = 'Tag';
}
