// @universe-design/react/src/breadcrumb/index.tsx
import type { HTMLAttributes, ReactElement } from 'react';
import React, { cloneElement, useState } from 'react';
import clsx from 'clsx';
import MoreOutlined from '@universe-design/icons-react/MoreOutlined';
import RightBoldOutlined from '@universe-design/icons-react/RightBoldOutlined';
import { useCssScope } from '../config-provider/hooks';
import type { BreadcrumbItemProps } from './BreadcrumbItem';
import { BreadcrumbItem } from './BreadcrumbItem';
import { Popover } from '../popover';
import { Overflow } from '../overflow';
import { toFlatArray } from '../util';

import './style';

export interface BreadcrumbProps extends HTMLAttributes<HTMLDivElement> {
  /**
   * 分隔符自定义
   * @defaultValue `<RightBoldOutlined />`
   */
  separator?: React.ReactNode;
  /**
   * 最多渲染的面包屑条目数量
   * @defaultValue Infinity
   */
  maxCount?: number | 'responsive';
}

export const Breadcrumb = Object.assign(
  React.forwardRef<HTMLDivElement, BreadcrumbProps>((props, ref) => {
    const { className, children, maxCount = Infinity, separator = <RightBoldOutlined />, ...restProps } = props;
    const cssScope = useCssScope('breadcrumb');
    const childNodes = toFlatArray(children) as ReactElement<BreadcrumbItemProps>[];
    const lastChildNode = childNodes.pop();

    const separatorNode = <span className={cssScope('-separator')}>{separator}</span>;
    const [visible, setVisible] = useState(false);

    return (
      <div className={clsx(cssScope(), className)} ref={ref} {...restProps}>
        <Overflow
          restPosition={1}
          maxCount={typeof maxCount === 'number' ? maxCount - 1 : 'responsive'}
          keys={Array(childNodes.length)
            .fill(0)
            .map((_, i) => i)}
          itemRenderer={key => (
            <>
              {childNodes[key]}
              {separatorNode}
            </>
          )}
          restRenderer={omittedKeys => (
            <>
              <Popover
                mode="pure"
                visible={visible}
                content={
                  <div onClick={() => setVisible(false)} className={cssScope('__popover__content')}>
                    {omittedKeys.map((o, i) => (
                      <React.Fragment key={o}>
                        {childNodes[o]}
                        {i !== omittedKeys.length - 1 && separatorNode}
                      </React.Fragment>
                    ))}
                  </div>
                }
                placement="bottomLeft"
                hideArrow
                onVisibleChange={setVisible}
              >
                <BreadcrumbItem className={cssScope('-ellipses')}>
                  <MoreOutlined />
                </BreadcrumbItem>
              </Popover>
              {separatorNode}
            </>
          )}
          suffixRenderer={() => (lastChildNode ? cloneElement(lastChildNode, { interactive: false }) : null)}
        />
      </div>
    );
  }),
  {
    Item: BreadcrumbItem,
  },
);
if (__DEV__) {
  Breadcrumb.displayName = 'Breadcrumb';
}

export type { BreadcrumbItemProps };
