// @universe-design/react/src/steps/stories/homeless.stories.tsx
/* eslint-disable max-lines-per-function */
import type { ReactNode } from 'react';
import React, { useState } from 'react';
import type { Meta } from '@storybook/react';
import { Steps } from '..';
import SendOutlined from '@universe-design/icons-react/SendOutlined';
import ThumbsupOutlined from '@universe-design/icons-react/ThumbsupOutlined';
import TimeOutlined from '@universe-design/icons-react/TimeOutlined';
import { Popover } from '../../popover';
import type { DotInfo } from '../interface';
export { default as VerticalWithIcon } from '../demo/vertical-icon.screenshot';
export { default as VerticalWithDot } from '../demo/vertical-dot.screenshot';

const { Step } = Steps;

export default {
  title: 'Components/Steps/临时示例', // 注意：这里需要修改成自己对应 Component name。
  component: Steps,
} satisfies Meta;

const stepStyle = { maxWidth: 600, marginBottom: 20 };
const verticalStepStyle = { maxWidth: 300, marginBottom: 20 };

export const AllSteps = () => {
  const [current, setCurrent] = useState(0);
  function handleChange(next: number) {
    console.log('change');
    setCurrent(next);
  }
  function progressDot(iconDot: ReactNode, { index, status }: DotInfo) {
    return (
      <Popover
        content={
          <span>
            step {index + 1} status: {status}
          </span>
        }
      >
        {iconDot}
      </Popover>
    );
  }

  return (
    <div>
      <h4>基本用法</h4>
      <Steps current={1} style={stepStyle}>
        <Step key="1" title="Finished" subTitle="Subtitle" description="This is a description." />
        <Step key="2" title="In Progress" description="This is a description." />
        <Step key="3" title="Waiting" description="This is a description." />
      </Steps>
      <Steps current={1} status="error" style={stepStyle}>
        <Step key="1" title="超长文案超长文案超长文案超长文案超长文案超长文案" description="This is a description." />
        <Step key="2" title="超长文案超长文案超长文案超长文案超长文案超长文案" description="This is a description." />
        <Step key="3" title="Waiting" description="This is a description." />
      </Steps>
      <h4>标题描述在图标右侧</h4>
      <Steps current={1} labelPlacement="horizontal" style={stepStyle}>
        <Step key="1" title="Finished" description="This is a description." />
        <Step key="2" title="In Progress" description="This is a description." />
        <Step key="3" title="Waiting" description="This is a description." />
      </Steps>
      <Steps current={1} labelPlacement="horizontal" status="error" style={stepStyle}>
        <Step key="1" title="Finished" description="This is a description." />
        <Step
          key="2"
          title="测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案"
          description="This is a description."
        />
        <Step key="3" title="Waiting" description="This is a description." />
      </Steps>
      <h4>竖直方向步骤条</h4>
      <Steps style={verticalStepStyle} current={1} direction="vertical">
        <Step key="1" title="Finished" description="This is a description." />
        <Step key="2" title="In Progress" description="This is a description." />
        <Step key="3" title="Waiting" description="This is a description." />
      </Steps>
      <Steps style={verticalStepStyle} current={1} direction="vertical" status="error">
        <Step
          key="1"
          title="测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案"
          description="This is a description.This is a description.This is a description.This is a description.This is a description."
        />
        <Step key="2" title="In Progress" description="This is a description." />
        <Step key="3" title="Waiting" description="This is a description." />
      </Steps>
      <Steps style={verticalStepStyle} current={1} direction="vertical">
        <Step key="1" title="Finished" />
        <Step key="2" title="In Progress" />
        <Step key="3" title="Waiting" />
      </Steps>
      <h4>点状水平步骤条</h4>
      <Steps style={stepStyle} current={current} type="dot" onChange={handleChange}>
        <Step key="1" title="Finished" description="This is a description." />
        <Step key="2" title="In Progress" description="This is a description." />
        <Step key="3" title="Waiting" description="This is a description." />
      </Steps>
      <Steps style={stepStyle} current={1} type="dot" status="error">
        <Step
          key="1"
          title="测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案"
          description="This is a description."
        />
        <Step
          key="2"
          title="测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案"
          description="This is a description.This is a description.This is a description.This is a description.This is a description."
        />
        <Step key="3" title="Waiting" description="This is a description." />
      </Steps>
      <h4>点状竖直步骤条</h4>
      <Steps style={verticalStepStyle} current={1} direction="vertical" type="dot">
        <Step key="1" title="Finished" description="This is a description." />
        <Step key="2" title="In Progress" description="This is a description." />
        <Step key="3" title="Waiting" description="This is a description." />
      </Steps>
      <Steps style={verticalStepStyle} current={1} direction="vertical" type="dot">
        <Step key="1" title="Finished" />
        <Step key="2" title="In Progress" />
        <Step key="3" title="Waiting" />
      </Steps>
      <Steps style={verticalStepStyle} current={1} direction="vertical" type="dot" status="error">
        <Step key="1" title="Finished" description="This is a description." />
        <Step
          key="2"
          title="测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案"
          description="This is a description.This is a description.This is a description.This is a description.This is a description."
        />
        <Step key="3" title="Waiting" description="This is a description." />
      </Steps>
      <h4>自定义图标步骤条</h4>
      <Steps>
        <Step status="finish" title="Login" icon={<SendOutlined />} />
        <Step status="process" title="Verification" icon={<TimeOutlined />} />
        <Step status="wait" title="Done" icon={<ThumbsupOutlined />} />
      </Steps>
      <h4>导航型步骤条</h4>
      <h5>无图标</h5>
      <Steps current={1} type="navigation" showIcon={false} style={stepStyle}>
        <Step key="1" title="Finished" description="This is a description." />
        <Step key="2" title="In Progress" description="This is a description." />
        <Step key="3" title="Waiting" description="This is a description." />
      </Steps>
      <Steps current={1} type="navigation" showIcon={false} style={stepStyle}>
        <Step key="1" title="Finished" />
        <Step key="2" title="In Progress" />
        <Step key="3" title="Waiting" />
      </Steps>
      <Steps current={1} type="navigation" status="error" style={stepStyle}>
        <Step
          key="1"
          title="测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案"
          description="测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案"
        />
        <Step key="2" title="In Progress" description="This is a description." />
        <Step key="3" title="Waiting" description="This is a description." />
      </Steps>
      <h5>有图标</h5>
      <Steps current={1} type="navigation" showIcon={true} style={stepStyle}>
        <Step key="1" title="Finished" description="This is a description." />
        <Step key="2" title="In Progress" description="This is a description." />
        <Step key="3" title="Waiting" description="This is a description." />
      </Steps>
      <Steps current={1} type="navigation" showIcon={true} style={stepStyle} status="error">
        <Step
          key="1"
          title="测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案"
          description="测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案"
        />
        <Step key="2" title="In Progress" description="This is a description." />
        <Step key="3" title="Waiting" description="This is a description." />
      </Steps>
      <h5>带下划线步骤条</h5>
      <Steps current={1} type="navigation" showNavIndicator={true} style={stepStyle}>
        <Step
          key="1"
          title="测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案"
          description="测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案"
        />
        <Step key="2" title="In Progress" description="This is a description." />
        <Step key="3" title="Waiting" description="This is a description." />
      </Steps>
      <h5>带图标下划线步骤条</h5>
      <Steps current={1} type="navigation" showIcon={true} showNavIndicator={true} style={stepStyle}>
        <Step key="1" title="Finished" description="This is a description." />
        <Step key="2" title="In Progress" description="This is a description." />
        <Step key="3" title="Waiting" description="This is a description." />
      </Steps>
      <Steps current={1} type="navigation" status="error" showIcon={true} showNavIndicator={true} style={stepStyle}>
        <Step
          key="1"
          title="测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案"
          description="测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案"
        />
        <Step key="2" title="In Progress" description="This is a description." />
        <Step key="3" title="Waiting" description="This is a description." />
      </Steps>
      <h5>可点击步骤条</h5>
      <Steps
        current={current}
        type="navigation"
        showIcon={true}
        showNavIndicator={true}
        style={stepStyle}
        onChange={handleChange}
      >
        <Step
          key="1"
          title="测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案"
          description="测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案"
        />
        <Step key="2" title="In Progress" description="This is a description." />
        <Step key="3" disabled title="Waiting" description="This is a description." />
      </Steps>
      <h5>tab</h5>
      <Steps current={current} type="navigation" showNavIndicator={true} style={stepStyle} onChange={handleChange}>
        <Step
          key="1"
          status="finish"
          title="测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案"
          description="测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案测试超长文案"
        />
        <Step key="2" status="process" title="In Progress" subTitle="subTitle" description="This is a description." />
        <Step key="3" status="wait" title="Waiting" description="This is a description." />
      </Steps>
      <h4>自定义节点</h4>
      <Steps current={1} style={stepStyle} progressDot={progressDot}>
        <Step key="1" title="Finished" subTitle="Subtitle" description="Hover on the dot." />
        <Step key="2" title="In Progress" description="Hover on the dot." />
        <Step key="3" title="Waiting" description="Hover on the dot." />
      </Steps>
      <Steps style={stepStyle} current={1} type="dot" progressDot={progressDot}>
        <Step key="1" title="Finished" description="Hover on the dot." />
        <Step key="2" title="In Progress" description="Hover on the dot." />
        <Step key="3" title="Waiting" description="Hover on the dot." />
      </Steps>
    </div>
  );
};
