// @universe-design/react/src/steps/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Basic from '../demo/0.basic';
import C1Icon from '../demo/1.icon';
import C2Change from '../demo/2.change';
import C3_1LabelPlacement from '../demo/3.1.label-placement';
import C3Vertical from '../demo/3.vertical';
import C4Error from '../demo/4.error';
import C5Dot from '../demo/5.dot';
import C6Clickable from '../demo/6.clickable';
import C7Nav from '../demo/7.nav';
import C8Custom from '../demo/8.custom';
if (!(C0Basic as Story).storyName) {
  (C0Basic as Story).storyName = '0. 基础用法';
  (C0Basic as Story).parameters = {
    docs: {
      description: {
        story: '简单的用法。\n\n',
      },
    },
  };
}
if (!(C1Icon as Story).storyName) {
  (C1Icon as Story).storyName = '1. 自定义图标步骤条';
  (C1Icon as Story).parameters = {
    docs: {
      description: {
        story: '自定义图标步骤条。\n\n',
      },
    },
  };
}
if (!(C2Change as Story).storyName) {
  (C2Change as Story).storyName = '2. 步骤切换';
  (C2Change as Story).parameters = {
    docs: {
      description: {
        story: '通常配合内容及按钮使用，表示一个流程的处理进度。\n\n',
      },
    },
  };
}
if (!(C3_1LabelPlacement as Story).storyName) {
  (C3_1LabelPlacement as Story).storyName = '3.1. 标签位置';
  (C3_1LabelPlacement as Story).parameters = {
    docs: {
      description: {
        story: '标签放置位置，默认水平放图标下面。\n\n',
      },
    },
  };
}
if (!(C3Vertical as Story).storyName) {
  (C3Vertical as Story).storyName = '3. 竖直方向步骤条';
  (C3Vertical as Story).parameters = {
    docs: {
      description: {
        story: '简单的竖直方向步骤条。\n\n',
      },
    },
  };
}
if (!(C4Error as Story).storyName) {
  (C4Error as Story).storyName = '4. 错误状态步骤条';
  (C4Error as Story).parameters = {
    docs: {
      description: {
        story: '使用 Steps 的 `status` 属性指定当前步骤状态。\n\n',
      },
    },
  };
}
if (!(C5Dot as Story).storyName) {
  (C5Dot as Story).storyName = '5. 点状步骤条';
  (C5Dot as Story).parameters = {
    docs: {
      description: {
        story: '点状步骤条。\n\n',
      },
    },
  };
}
if (!(C6Clickable as Story).storyName) {
  (C6Clickable as Story).storyName = '6. 可点击步骤条';
  (C6Clickable as Story).parameters = {
    docs: {
      description: {
        story: '设置 `onChange` 后，步骤条可点击。\n\n',
      },
    },
  };
}
if (!(C7Nav as Story).storyName) {
  (C7Nav as Story).storyName = '7. 导航步骤条';
  (C7Nav as Story).parameters = {
    docs: {
      description: {
        story: '导航类型步骤条。\n\n',
      },
    },
  };
}
if (!(C8Custom as Story).storyName) {
  (C8Custom as Story).storyName = '8. 自定义节点';
  (C8Custom as Story).parameters = {
    docs: {
      description: {
        story: '自定义节点步骤条。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Steps/官网示例',
  tags: ['autodocs'],
};

export { C0Basic, C1Icon, C2Change, C3_1LabelPlacement, C3Vertical, C4Error, C5Dot, C6Clickable, C7Nav, C8Custom };
