// @byted-service/bytedtrace/lib/sampler/limiter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LimiterPack = exports.newAtomicLimiterFunc = exports.TickLimiter = exports.AtomicLimiter = exports.SimpleLimiter = exports.RateLimiter = void 0;
const tslib_1 = require("tslib");
const gUtils = tslib_1.__importStar(require("../utils"));
const consts = tslib_1.__importStar(require("./consts"));
class RateLimiter {
    constructor() {
        this.updateBalanceInterval = 1000; // ms
        this.lastTick = 0;
    }
    setMaxCredit(max) {
        this.maxBalance = Number(max);
    }
    setMaxHitPerSecond(hitPerSecond) {
        this.creditsPerSecond = hitPerSecond;
        this.balance = Number(gUtils.rangeValue(1, hitPerSecond));
    }
    pass() {
        return this.checkCredit(1);
    }
    checkCredit(itemCost) {
        const currentTime = Date.now();
        const elapsedTime = currentTime - this.lastTick;
        if (elapsedTime >= this.updateBalanceInterval) {
            this.lastTick = currentTime;
            this.balance += elapsedTime * this.creditsPerSecond;
            if (this.balance > this.maxBalance) {
                this.balance = this.maxBalance;
            }
        }
        if (this.balance >= itemCost) {
            this.balance -= itemCost;
            return true;
        }
        return false;
    }
}
exports.RateLimiter = RateLimiter;
class SimpleLimiter {
    constructor() {
        this.lastHitMs = Date.now();
    }
    // There is no maxCredit of simple limiter.
    setMaxCredit(max) {
        // do nothing
    }
    // 设置每秒钟最大的命中数.
    setMaxHitPerSecond(hitPerSecond) {
        this.intervalMs = (1 / hitPerSecond) * 1000;
    }
    pass() {
        const nowTime = Date.now();
        if (nowTime - this.lastHitMs > this.intervalMs) {
            this.lastHitMs = nowTime;
            return true;
        }
        return false;
    }
}
exports.SimpleLimiter = SimpleLimiter;
// 思路与rateLimiter类似.
// 并发下性能更好.
class AtomicLimiter {
    constructor() {
        this.lastHit = Date.now();
    }
    setMaxCredit(max) {
        this.maxCredit = max;
    }
    setMaxHitPerSecond(hitPerSecond) {
        this.incPerMs = hitPerSecond / 1000;
        this.credit = gUtils.rangeValue(1, hitPerSecond);
    }
    pass() {
        if (this.credit >= 1) {
            this.credit--;
            return true;
        }
        const t = Date.now();
        const delta = (t - this.lastHit) * this.incPerMs;
        if (delta >= 1) {
            if (delta > this.maxCredit) {
                this.credit = this.maxCredit;
            }
            else {
                this.credit += delta;
            }
            this.lastHit = t;
        }
        if (this.credit >= 1) {
            this.credit--;
            return true;
        }
        else {
            return false;
        }
    }
}
exports.AtomicLimiter = AtomicLimiter;
// Record last hit time for expire.
class TickLimiter {
    constructor(newLimiter, expiredTimeMs) {
        this.limiter = newLimiter;
        this.hitInMs = Date.now();
        this.expiredTimeMs = expiredTimeMs;
    }
    setMaxCredit(max) {
        this.limiter.setMaxCredit(max);
    }
    setMaxHitPerSecond(hitPerSecond) {
        this.limiter.setMaxHitPerSecond(hitPerSecond);
    }
    pass() {
        if (this.expiredTimeMs >= 0) {
            this.hitInMs = Date.now();
        }
        return this.limiter.pass();
    }
    shouldExpire() {
        if (this.expiredTimeMs < 0) {
            return false;
        }
        return Date.now() - this.hitInMs > this.expiredTimeMs;
    }
}
exports.TickLimiter = TickLimiter;
function newAtomicLimiterFunc(maxHitPerSecond, maxCredit) {
    const mc = getMaxCredit(maxHitPerSecond, maxCredit);
    const limiter = new AtomicLimiter();
    limiter.setMaxHitPerSecond(maxHitPerSecond);
    limiter.setMaxCredit(mc);
    return limiter;
}
exports.newAtomicLimiterFunc = newAtomicLimiterFunc;
function getMaxCredit(maxHitPerSecond, maxCredit) {
    if (maxCredit < 0) {
        if (maxHitPerSecond <= 0) {
            return 0;
        }
        if (maxHitPerSecond < 1) {
            return 1;
        }
        else {
            return parseInt(maxHitPerSecond.toString(), 10);
        }
    }
    return maxCredit;
}
// Limit集合器.
class LimiterPack {
    constructor(maxOperationNum, expireTime, newLimiter) {
        this.limiters = new Map();
        this.maxOperationNum = maxOperationNum;
        this.newLimiter = newLimiter;
        this.expireTime = expireTime;
        this.lastApplyTime = 0;
    }
    pass(name) {
        const limiter = this.limiters.get(name);
        if (limiter) {
            return limiter.pass();
        }
        else {
            if (this.limiters.size > this.maxOperationNum && !this.applyOne()) {
                return false;
            }
            const newLimiter = new TickLimiter(this.newLimiter, this.expireTime);
            this.limiters.set(name, newLimiter);
            return newLimiter.pass();
        }
    }
    insertLimiter(name, limiter) {
        this.limiters.set(name, limiter);
    }
    /**
     * 如果 Date.now() - lastApplyTime 大于规定的间隔时间，
     * 则进行检查是否有过去的 TickLimiter，如果过期则删除
     * @returns
     */
    applyOne() {
        const interval = Date.now() - this.lastApplyTime;
        if (interval < consts.APPLY_ONE_INTERVAL_MS) {
            return false;
        }
        let hasOne = false;
        for (const [key, value] of this.limiters) {
            if (value.shouldExpire()) {
                this.limiters.delete(key);
                hasOne = true;
            }
        }
        this.lastApplyTime = Date.now();
        return hasOne;
    }
}
exports.LimiterPack = LimiterPack;
