// @byted-service/bytedtrace/lib/sampler/qpsCatcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QpsCatcher = void 0;
const tslib_1 = require("tslib");
const consts = tslib_1.__importStar(require("./consts"));
const counter_1 = require("./counter");
class QpsCatcher {
    constructor(isNoop = false, options = {}) {
        if (isNoop) {
            this.isNoop = isNoop;
            return this;
        }
        const duration = options.windowDuringMs * options.windowNum;
        let windowsTimeSpan = duration / 1000.0;
        /* istanbul ignore next */
        if (windowsTimeSpan === 0) {
            windowsTimeSpan = 1 / Number.MAX_VALUE;
        }
        this.maxOperationNum = options.maxOperationNum;
        this.windowDuringMs = options.windowDuringMs;
        this.windowNum = options.windowNum;
        this.windowExpireTimeMs = options.windowExpireTimeMs;
        this.windowCounterNum = 0;
        this.windowCounters = new Map();
        this.windowsTimeSpan = windowsTimeSpan;
        this.ticker = new Ticker(options.windowDuringMs, this.tickHandler.bind(this));
        this.tick();
    }
    // 打一个点,传入方法名.
    inc(operationName, delta) {
        if (this.isNoop) {
            return;
        }
        const counter = this.getWindowCounter(operationName);
        counter.add(delta);
    }
    // 打一个点同时返回qps
    incAndGetQps(operationName, delta) {
        if (this.isNoop) {
            return 0;
        }
        const counter = this.getWindowCounter(operationName);
        const ct = counter.getAndAdd(delta);
        return ct / this.windowsTimeSpan;
    }
    getQps(operationName) {
        if (this.isNoop) {
            return 0;
        }
        const counter = this.windowCounters.get(operationName);
        if (!counter) {
            return -1;
        }
        const ct = counter.get();
        return ct / this.windowsTimeSpan;
    }
    tickHandler() {
        for (const [key, value] of this.windowCounters) {
            if (value) {
                if (value.shouldExpired()) {
                    this.windowCounters.delete(key);
                    this.windowCounterNum -= 1;
                }
                else {
                    value.tick();
                }
            }
        }
    }
    tick() {
        this.ticker.tick();
    }
    close() {
        if (this.isNoop) {
            return;
        }
        this.ticker.stop();
    }
    getWindowCounter(operationName) {
        const counter = this.windowCounters.get(operationName);
        if (counter) {
            return counter;
        }
        return this.applyOne(operationName);
    }
    applyOne(operationName) {
        if (this.windowCounterNum >= this.maxOperationNum) {
            throw new Error(consts.ERROR_OVER_MAX_WINDOW_NUM);
        }
        const newCounter = new counter_1.WindowCounter(this.windowNum, this.windowExpireTimeMs / this.windowDuringMs);
        this.windowCounters.set(operationName, newCounter);
        this.windowCounterNum += 1;
        return newCounter;
    }
}
exports.QpsCatcher = QpsCatcher;
class Ticker {
    constructor(duration, callback) {
        this.duration = duration;
        this.callback = callback;
    }
    tick() {
        if (!this.id) {
            this.id = setInterval(this.callback, this.duration).unref();
        }
    }
    stop() {
        if (this.id) {
            clearInterval(this.id);
            this.id = undefined;
        }
    }
}
