// @universe-design/react/src/table/hooks/useFilter/FilterMenu.tsx
import React from 'react';
import clsx from 'clsx';
import { Menu } from '../../../menu';
import { Button } from '../../../button';
import { Checkbox } from '../../../checkbox';
import { Radio } from '../../../radio';
import { Empty } from '../../../empty';
import type { TableLocale, ColumnFilterItem, Key, GetPopupContainerInTable } from '../../interface';
import { serializeFilterKey } from '../../utils';
import { ListDivider, ListItemLike } from '../../../shared/components/list';

function hasSubMenu(filters: ColumnFilterItem[]) {
  return filters.some(({ children }) => children);
}

interface FilterMenuProps {
  prefixCls: string;
  dropdownPrefixCls: string;
  tableLocale: TableLocale;
  filters: ColumnFilterItem[];
  filterMultiple: boolean;
  openKeys: Key[];
  selectedKeys: Key[];
  getPopupContainer?: GetPopupContainerInTable;
  onMenuClick: () => void;
  onSelectKeys: ({ selectedKeys }: { selectedKeys: Key[] }) => void;
  onOpenChange: (openKeys: Key[]) => void;
  onReset: () => void;
  onConfirm: () => void;
}

export default function FilterMenu(props: FilterMenuProps) {
  const {
    prefixCls,
    dropdownPrefixCls,
    tableLocale,
    filters,
    filterMultiple,
    openKeys,
    selectedKeys,
    onMenuClick,
    onSelectKeys,
    onOpenChange,
    onReset,
    onConfirm,
    getPopupContainer,
  } = props;

  const renderFilterItems = (colFilters: ColumnFilterItem[]) => {
    if (colFilters.length === 0) {
      return (
        <div
          style={{
            margin: '12px 0',
          }}
        >
          <Empty
            description={tableLocale.filterEmptyText}
            imageStyle={{
              height: 36,
            }}
          />
        </div>
      );
    }
    return colFilters.map((filter, index) => {
      const key = serializeFilterKey(filter.value);

      if (filter.children) {
        return (
          <Menu.SubMenu key={key || index} title={filter.text} popupClassName={`${prefixCls}-dropdown-submenu`}>
            {renderFilterItems(filter.children)}
          </Menu.SubMenu>
        );
      }

      const Component = filterMultiple ? Checkbox : Radio;

      return (
        <Menu.Item key={filter.value !== undefined ? key : index}>
          <Component checked={selectedKeys.includes(key)} />
          <span className={`${prefixCls}-dropdown-label`}>{filter.text}</span>
        </Menu.Item>
      );
    });
  };

  const triggerSelect = (pramas: { selectedKeys?: Key[] }) => {
    onSelectKeys({
      selectedKeys: pramas.selectedKeys ?? [],
    });
  };

  return (
    <>
      <Menu
        multiple={filterMultiple}
        // prefixCls={`${dropdownPrefixCls}__menu`}
        className={clsx({
          [`${dropdownPrefixCls}__menu-without-submenu`]: !hasSubMenu(filters),
        })}
        mode="normal"
        onClick={onMenuClick}
        onSelect={triggerSelect}
        onDeselect={triggerSelect}
        selectedKeys={selectedKeys as string[]}
        getPopupContainer={getPopupContainer}
        openKeys={openKeys as string[]}
        onOpenChange={onOpenChange}
        selectable={true}
        showCheckmark={false}
      >
        {renderFilterItems(filters)}
      </Menu>
      <ListDivider fluid />
      <ListItemLike className={`${prefixCls}-dropdown-btns`}>
        <Button type="link" size="small" disabled={selectedKeys.length === 0} onClick={onReset}>
          {tableLocale.filterReset}
        </Button>
        <Button type="primary" size="small" onClick={onConfirm}>
          {tableLocale.filterConfirm}
        </Button>
      </ListItemLike>
    </>
  );
}
