// @universe-design/react/src/table/hooks/useFilter/FilterDropdown.tsx
import React, { useRef, useMemo, useCallback, useState, useEffect } from 'react';
import clsx from 'clsx';
import isEqual from 'lodash/isEqual';
import FilterOutlined from '@universe-design/icons-react/FilterOutlined';
import { Dropdown } from '../../../dropdown';
import type { ColumnType, Key, GetPopupContainerInTable, TableLocale, FilterState, FilterKey } from '../../interface';
import { useForceUpdate } from '@universe-design/react-utilities';
import { isFunction } from '../../../util';
import FilterMenu from './FilterMenu';
import { IconButton } from '../../../button';
import { serializeFilterKeys, unSerializeFilterKeys } from '../../utils';
import { ListBox } from '../../../shared/components/list/List';

export interface FilterDropdownProps<RecordType> {
  prefixCls: string;
  dropdownPrefixCls: string;
  tableLocale: TableLocale;
  column: ColumnType<RecordType>;
  filterState?: FilterState<RecordType>;
  filterMultiple: boolean;
  columnKey: Key;
  children: React.ReactNode;
  triggerFilter: (filterState: FilterState<RecordType>) => void;
  getPopupContainer?: GetPopupContainerInTable;
}

function FilterDropdown<RecordType>(props: FilterDropdownProps<RecordType>) {
  const {
    column,
    prefixCls,
    dropdownPrefixCls,
    tableLocale,
    columnKey,
    filterState,
    triggerFilter,
    children,
    filterMultiple,
    getPopupContainer,
  } = props;

  const { filterDropdownVisible, onFilterDropdownVisibleChange } = column;
  const [visible, setVisible] = React.useState(false);

  const clickMenuRef = useRef(false);

  const forceUpdate = useForceUpdate();

  const filtered = Boolean(filterState && (filterState.filteredKeys?.length || filterState.forceFiltered));

  // dropdown visible
  const mergedVisible = typeof filterDropdownVisible === 'boolean' ? filterDropdownVisible : visible;
  const triggerVisible = (newVisible: boolean) => {
    setVisible(newVisible);
    if (onFilterDropdownVisibleChange) {
      onFilterDropdownVisibleChange(newVisible);
    }
  };

  // Select Keys
  const propFilteredKeys = useMemo(() => filterState?.filteredKeys || [], [filterState]);

  const filteredKeysSync = useRef(propFilteredKeys);
  const getFilteredKeysSync = () => filteredKeysSync.current;
  const setFilteredKeysSync = useCallback(
    (value: FilterKey[]) => {
      filteredKeysSync.current = value;
      forceUpdate();
    },
    [filteredKeysSync, forceUpdate],
  );

  const onSelectKeys = useCallback(
    (param: { selectedKeys: Key[] }) => {
      setFilteredKeysSync(unSerializeFilterKeys(param.selectedKeys as string[]));
    },
    [setFilteredKeysSync],
  );

  useEffect(() => {
    onSelectKeys({ selectedKeys: serializeFilterKeys(propFilteredKeys) || [] });
  }, [propFilteredKeys, onSelectKeys]);

  // Open Keys
  const [openKeys, setOpenKeys] = useState<Key[]>([]);

  // Open
  const onOpenChange = (keys: Key[]) => {
    setOpenKeys(keys);
  };

  const onMenuClick = () => {
    clickMenuRef.current = true;
  };

  // Submit
  const internalTriggerFilter = (keys: FilterKey[] | undefined | null) => {
    triggerVisible(false);

    const mergedKeys = keys && keys.length ? keys : null;
    if (mergedKeys === null && (!filterState || !filterState.filteredKeys)) {
      return null;
    }

    if (isEqual(mergedKeys, filterState?.filteredKeys)) {
      return null;
    }

    triggerFilter({
      column,
      key: columnKey,
      filteredKeys: mergedKeys,
    });
  };

  const onConfirm = () => {
    internalTriggerFilter(getFilteredKeysSync());
  };

  const onReset = () => {
    setFilteredKeysSync([]);
    internalTriggerFilter([]);
  };

  const onVisibleChange = (newVisible: boolean) => {
    if (newVisible && propFilteredKeys !== undefined) {
      setFilteredKeysSync(propFilteredKeys);
    }
    if (!clickMenuRef.current) {
      triggerVisible(newVisible);
    }
    if (!clickMenuRef.current && !newVisible && !column.filterDropdown) {
      onConfirm();
    }
    clickMenuRef.current = false;
  };

  let dropdownContent: React.ReactNode, filterIconComp: React.ReactNode;

  if (isFunction(column.filterDropdown)) {
    const selectedKeys = getFilteredKeysSync() || [];
    dropdownContent = column.filterDropdown({
      prefixCls: `${dropdownPrefixCls}-custom`,
      setSelectedKeys: (keys: FilterKey[]) => onSelectKeys({ selectedKeys: serializeFilterKeys(keys) }),
      selectedKeys,
      confirm: onConfirm,
      clearFilters: onReset,
      filters: column.filters,
      visible: mergedVisible,
    });
  } else if (column.filterDropdown) {
    dropdownContent = column.filterDropdown;
  } else {
    const selectedKeys = serializeFilterKeys(getFilteredKeysSync() || []);
    dropdownContent = (
      <FilterMenu
        prefixCls={`${prefixCls}-filter`}
        tableLocale={tableLocale}
        dropdownPrefixCls={dropdownPrefixCls}
        filterMultiple={filterMultiple}
        filters={column.filters || []}
        openKeys={openKeys}
        selectedKeys={selectedKeys}
        onMenuClick={onMenuClick}
        onSelectKeys={onSelectKeys}
        onOpenChange={onOpenChange}
        onConfirm={onConfirm}
        onReset={onReset}
        getPopupContainer={getPopupContainer}
      />
    );
  }

  if (isFunction(column.filterIcon)) {
    filterIconComp = column.filterIcon(filtered);
  } else if (column.filterIcon) {
    filterIconComp = column.filterIcon;
  } else {
    filterIconComp = <FilterOutlined />;
  }

  const menu = <ListBox>{dropdownContent}</ListBox>;

  const dropDownNode = (
    <Dropdown
      overlay={menu}
      trigger={['click']}
      visible={mergedVisible}
      onVisibleChange={onVisibleChange}
      getPopupContainer={getPopupContainer}
      overlayClassName={`${prefixCls}-filter-dropdown`}
    >
      <IconButton
        color={filtered ? 'primary' : undefined}
        selected={filtered || mergedVisible}
        icon={filterIconComp}
        size="small"
      />
    </Dropdown>
  );

  return (
    <>
      {children}
      <span
        className={clsx(`${prefixCls}-column-filter`, {
          [`${prefixCls}-filter-trigger-container-open`]: mergedVisible,
        })}
        onClick={event => {
          event.stopPropagation();
        }}
      >
        {dropDownNode}
      </span>
    </>
  );
}

export default FilterDropdown;
