// @universe-design/react/src/table/hooks/useFilter/index.tsx
import React, { useState, useMemo, useCallback } from 'react';
import FilterDropdown from './FilterDropdown';
import { getColumnPos, renderColumnTitle, getColumnKey } from '../../utils';
import type {
  TransformColumns,
  ColumnsType,
  ColumnType,
  ColumnTitleProps,
  Key,
  GetPopupContainerInTable,
  ColumnFilterItem,
  TableLocale,
  FilterState,
  FilterKey,
} from '../../interface';
import clsx from 'clsx';

function collectFilterStates<RecordType>(
  columns: ColumnsType<RecordType>,
  init: boolean,
  pos?: string,
): FilterState<RecordType>[] {
  let filterStates: FilterState<RecordType>[] = [];

  (columns || []).forEach((column, index) => {
    const columnPos = getColumnPos(index, pos);

    if ('children' in column) {
      filterStates = [...filterStates, ...collectFilterStates(column.children, init, columnPos)];
    } else if (column.filters || 'filterDropdown' in column || 'onFilter' in column) {
      if ('filteredValue' in column) {
        filterStates.push({
          column,
          key: getColumnKey(column, columnPos),
          filteredKeys: column.filteredValue || [],
          forceFiltered: column.filtered,
        });
      } else {
        filterStates.push({
          column,
          key: getColumnKey(column, columnPos),
          filteredKeys: init && column.defaultFilteredValue ? column.defaultFilteredValue : undefined,
          forceFiltered: column.filtered,
        });
      }
    }
  });

  return filterStates;
}

function generateFilterInfo<RecordType>(filterStates: FilterState<RecordType>[]) {
  const currentFilters: Record<string, (Key | boolean)[] | null> = {};

  filterStates.forEach(({ key, filteredKeys, column }) => {
    const { filters, filterDropdown } = column;
    if (filterDropdown) {
      currentFilters[key] = filteredKeys || null;
    } else if (Array.isArray(filteredKeys)) {
      const keys = flattenKeys(filters);
      currentFilters[key] = keys.filter(originKey => filteredKeys.includes(originKey));
    } else {
      currentFilters[key] = null;
    }
  });

  return currentFilters;
}

function flattenKeys(filters?: ColumnFilterItem[]) {
  let keys: FilterKey[] = [];
  (filters || []).forEach(({ value, children }) => {
    keys.push(value);
    if (children) {
      keys = [...keys, ...flattenKeys(children)];
    }
  });
  return keys;
}

export function getFilterData<RecordType>(data: readonly RecordType[], filterStates: FilterState<RecordType>[]) {
  return filterStates.reduce((currentData, filterState) => {
    const {
      column: { onFilter, filters },
      filteredKeys,
    } = filterState;
    if (onFilter && filteredKeys && filteredKeys.length) {
      return currentData.filter(record =>
        filteredKeys.some(key => {
          const keys = flattenKeys(filters);
          const keyIndex = keys.findIndex(k => k === key);
          const realKey = keyIndex !== -1 ? keys[keyIndex] : key;
          return onFilter(realKey, record);
        }),
      );
    }
    return currentData;
  }, data);
}

interface FilterConfig<RecordType> {
  prefix: string;
  dropdownPrefix: string;
  tableLocale: TableLocale;
  mergedColumns: ColumnsType<RecordType>;
  onFilterChange: (filters: Record<string, FilterKey[] | null>, filterStates: FilterState<RecordType>[]) => void;
  getPopupContainer?: GetPopupContainerInTable;
}

function useFilter<RecordType>({
  prefix,
  dropdownPrefix,
  tableLocale,
  mergedColumns,
  onFilterChange,
  getPopupContainer,
}: FilterConfig<RecordType>): [
  TransformColumns<RecordType>,
  FilterState<RecordType>[],
  () => Record<string, FilterKey[] | null>,
] {
  const [filterStates, setFilterStates] = useState<FilterState<RecordType>[]>(collectFilterStates(mergedColumns, true));

  const mergedFilterStates = useMemo(() => {
    const collectedStates = collectFilterStates(mergedColumns, false);
    if (collectedStates.every(({ filteredKeys }) => filteredKeys === undefined)) {
      return filterStates;
    }
    return collectedStates;
  }, [mergedColumns, filterStates]);

  const triggerFilter = useCallback(
    (filterState: FilterState<RecordType>) => {
      const newFilterStates = mergedFilterStates.filter(({ key }) => key !== filterState.key);
      newFilterStates.push(filterState);
      setFilterStates(newFilterStates);
      onFilterChange(generateFilterInfo(newFilterStates), newFilterStates);
    },
    [mergedFilterStates, setFilterStates, onFilterChange],
  );

  const transformColumns = useCallback(
    (innerColumns: ColumnsType<RecordType>): ColumnsType<RecordType> => {
      const injectFilter = (
        columns: ColumnsType<RecordType>,
        states: FilterState<RecordType>[],
        pos?: string,
      ): ColumnsType<RecordType> =>
        columns.map((column, index) => {
          const columnPos = getColumnPos(index, pos);
          const { filterMultiple = true } = column as ColumnType<RecordType>;

          let newColumn: ColumnsType<RecordType>[number] = column;

          if (newColumn.filters || newColumn.filterDropdown) {
            const columnKey = getColumnKey(newColumn, columnPos);
            const filterState = states.find(({ key }) => columnKey === key);

            newColumn = {
              ...newColumn,
              className: clsx(newColumn.className, {
                [`${prefix}-column-filter-align-right`]: newColumn.filterIconAlign === 'right',
              }),
              title: (renderProps: ColumnTitleProps<RecordType>) => (
                <FilterDropdown
                  prefixCls={prefix}
                  dropdownPrefixCls={dropdownPrefix}
                  tableLocale={tableLocale}
                  column={newColumn}
                  columnKey={columnKey}
                  filterState={filterState}
                  filterMultiple={filterMultiple}
                  triggerFilter={triggerFilter}
                  getPopupContainer={getPopupContainer}
                >
                  {renderColumnTitle(column.title, renderProps)}
                </FilterDropdown>
              ),
              onHeaderCell: col => {
                const cell: React.HTMLAttributes<HTMLElement> = (column.onHeaderCell && column.onHeaderCell(col)) || {};
                cell.className = clsx(cell.className, `${prefix}-column-has-filters`);

                return cell;
              },
            };
          }

          if ('children' in newColumn) {
            newColumn = {
              ...newColumn,
              children: injectFilter(newColumn.children, states, columnPos),
            };
          }

          return newColumn;
        });

      return injectFilter(innerColumns, mergedFilterStates);
    },
    [prefix, dropdownPrefix, tableLocale, getPopupContainer, triggerFilter, mergedFilterStates],
  );

  const getFilters = useCallback(() => generateFilterInfo(mergedFilterStates), [mergedFilterStates]);

  return [transformColumns, mergedFilterStates, getFilters];
}

export default useFilter;
