// @universe-design/react/es/form/FormItemUI.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.u1e8zt0z{min-width:min-content;}
import u10wqw00 from "./../.stylex/u10wqw00";
import InfoOutlined from '@universe-design/icons-react/InfoOutlined';
import clsx from 'clsx';
import React from 'react';
import stylex from '@universe-design/stylex';
import { Col } from '../col';
import { Row } from '../row';
import { useCssScope } from '../config-provider/hooks';
import { Tooltip } from '../tooltip';
import { isFunction, isNonNullable } from '../util';
import { useFormUIContextSelector } from './FormUIContext';
import { forwardHTMLRef } from '../util/forwardRef';
import { Text } from '../text';
import './style';
/**
 * @public
 */

/**
 * @public
 */

/**
 * @public
 * @docgen
 */

export const FormItemUI = forwardHTMLRef((props, ref) => {
  var _ref, _ref2, _ref3, _ref4;
  const {
    label,
    className,
    style,
    children,
    labelAlign,
    labelCol,
    required = false,
    wrapperCol,
    help,
    hidden,
    validateStatus,
    tooltip,
    extra,
    htmlFor,
    labelInfoIconRenderer,
    errorDisplay: errorDisplayProp,
    ...restProps
  } = props;
  const formCtxLabelAlign = useFormUIContextSelector(ctx => ctx.labelAlign);
  const formCtxLabelCol = useFormUIContextSelector(ctx => ctx.labelCol);
  const formCtxWrapperCol = useFormUIContextSelector(ctx => ctx.wrapperCol);
  const formCtxErrorDisplay = useFormUIContextSelector(ctx => ctx.errorDisplay);
  const mergedLabelAlign = (_ref = labelAlign !== null && labelAlign !== void 0 ? labelAlign : formCtxLabelAlign) !== null && _ref !== void 0 ? _ref : 'right';
  const mergedLabelCol = (_ref2 = labelCol !== null && labelCol !== void 0 ? labelCol : formCtxLabelCol) !== null && _ref2 !== void 0 ? _ref2 : {};
  const mergedWrapperCol = (_ref3 = wrapperCol !== null && wrapperCol !== void 0 ? wrapperCol : formCtxWrapperCol) !== null && _ref3 !== void 0 ? _ref3 : {};
  const errorDisplay = (_ref4 = errorDisplayProp !== null && errorDisplayProp !== void 0 ? errorDisplayProp : formCtxErrorDisplay) !== null && _ref4 !== void 0 ? _ref4 : 'locally';
  const itemCssScope = useCssScope('form__item');
  const itemLabelCssScope = useCssScope('form__item__label');
  const itemControlCssScope = useCssScope('form__item__control');
  const hasExplain = isNonNullable(help) && errorDisplay !== 'none';
  /**
   * ================ Tooltip Content Node ================
   */
  const {
    tooltipProps,
    labelInfoIconNode
  } = (() => {
    let _tooltipProps = {};
    if (tooltip && typeof tooltip === 'object' && ! /*#__PURE__*/React.isValidElement(tooltip)) {
      _tooltipProps = tooltip;
    } else {
      _tooltipProps = {
        title: tooltip
      };
    }
    const {
      icon = /*#__PURE__*/React.createElement(InfoOutlined, null),
      ...restTooltipProps
    } = _tooltipProps;
    const _labelInfoIconNode = /*#__PURE__*/React.createElement(Text, {
      as: "div",
      className: itemCssScope('__tooltip')
    }, icon);
    return {
      tooltipProps: restTooltipProps,
      labelInfoIconNode: _labelInfoIconNode
    };
  })();
  const tooltipNode = (() => {
    if (isFunction(labelInfoIconRenderer)) {
      return labelInfoIconRenderer(labelInfoIconNode);
    }
    if (tooltip) {
      return /*#__PURE__*/React.createElement(Tooltip, tooltipProps, labelInfoIconNode);
    }
    return null;
  })();

  /**
   * ================ Extra Node ================
   */
  const extraNode = extra ? /*#__PURE__*/React.createElement(Text, {
    className: itemControlCssScope('__extra')
  }, extra) : null;

  /**
   * ================ Error Or Help Node ================
   */
  const errorOrHelpNode = hasExplain ? /*#__PURE__*/React.createElement(Text, {
    className: clsx(itemControlCssScope('__explain'), validateStatus === 'error' && itemControlCssScope('__explain-error'))
  }, help) : null;
  return /*#__PURE__*/React.createElement(Row, _extends({}, restProps, {
    className: clsx(itemCssScope(), hasExplain && itemCssScope('-with-help'), hidden && itemCssScope('-hidden'), itemCssScope(`-error-${errorDisplay}`), className),
    style: style,
    ref: ref
  }), isNonNullable(label) ? /*#__PURE__*/React.createElement(Col, _extends({}, mergedLabelCol, {
    className: clsx(itemLabelCssScope(), mergedLabelAlign === 'left' && itemLabelCssScope('-left'), mergedLabelCol.className)
  }), /*#__PURE__*/React.createElement(Text, {
    as: "label",
    htmlFor: htmlFor,
    title: typeof label === 'string' ? label : undefined,
    lineClamp: 2,
    sx: [{
      "props": {
        "Bf4jedk": u10wqw00
      }
    }]
  }, label), required && /*#__PURE__*/React.createElement(Text, {
    className: itemCssScope('__required-mark')
  }, "*"), tooltipNode) : null, /*#__PURE__*/React.createElement(Col, _extends({}, mergedWrapperCol, {
    className: clsx(itemControlCssScope(), mergedWrapperCol.className)
  }), /*#__PURE__*/React.createElement("div", {
    className: itemControlCssScope('__input')
  }, /*#__PURE__*/React.createElement("div", {
    className: itemControlCssScope('__input__content')
  }, children)), errorOrHelpNode, extraNode));
});
if (process.env.NODE_ENV !== "production") {
  FormItemUI.displayName = 'FormItemUI';
}
