// @universe-design/react/es/form/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
/* eslint-disable react/destructuring-assignment */
import clsx from 'clsx';
import React, { isValidElement, forwardRef, memo, useMemo, useImperativeHandle } from 'react';
import merge from 'lodash/merge';
import get from 'lodash/get';
import scrollIntoView from 'scroll-into-view-if-needed';
import { Field, useFieldContext, marshalNamePath, FieldPropsAreEqual, FormList as MiniFormList, useForm, FormContextProvider } from '../react-mini-form';
import { useComponentSize, useCssScope, useLocaleReceiver } from '../config-provider/hooks';
import { devWarning, isFunction, isNonNullable, getDefaultSizeAbbr, warning } from '../util';
import { validateRule } from './utils/validateUtil';
import { FormUIContextProvider, useFormUIContextSelector } from './FormUIContext';
import { FormItemUI } from './FormItemUI';
import { unmarshalNamePath, asArray } from '../react-mini-form/shared';
import { ConfigProvider } from '../config-provider';
import { FindDOMNode } from '@universe-design/react-utilities';
import { FormDescendantsProvider, useFormDescendant, useFormDescendants } from './FormDescendantContext';
import './style';

/**
 * @public
 */

/**
 * @public
 * @docgen
 */

/**
 * @docgen
 * @public
 * 表单字段组件`Form.Item`，用于数据双向绑定、校验、布局等。
 */

/**
 * @docgen
 * @public
 * @typeParam Values - Values of this form
 */

/**
 * @public
 */

/**
 * @docgen
 * @public
 * @typeParam FieldValue - FieldValue of this form list
 */

const BaseForm = /*#__PURE__*/forwardRef(function (props, ref) {
  const size = useComponentSize();
  const {
    layout = 'vertical',
    className,
    requiredMark = true,
    size: sizeProp = size,
    labelAlign,
    labelCol,
    wrapperCol,
    onFinish,
    validateMessages,
    errorDisplay,
    component = 'form',
    children,
    shouldScrollToFirstErrorField = true,
    onSubmitError,
    ...restProps
  } = props;

  /**
   * ================ ClassName ================
   */
  const formCssScope = useCssScope('form');
  const formClassName = clsx(formCssScope(), formCssScope(`-${layout}`), !requiredMark && formCssScope('-hide-required-mark'), formCssScope(`-${getDefaultSizeAbbr(sizeProp, 'medium')}`), className);

  /**
   * ================ Logic context ================
   */
  const formCtx = useForm({
    ...restProps,
    onSubmit: onFinish,
    onSubmitError: errors => {
      isFunction(onSubmitError) && onSubmitError(errors);
      // eslint-disable-next-line @typescript-eslint/no-use-before-define
      scrollToFirstError();
    }
  });
  const {
    submit
  } = formCtx;

  /**
   * ================ Descendants ================
   */
  const descendants = useFormDescendants();

  /**
   * ================ Methods ================
   */

  // ---------------- Handle submit ----------------
  const handleSubmit = React.useCallback(e => {
    e.preventDefault();
    e.stopPropagation();
    submit();
  }, [submit]);
  let childNode = children;
  if (component) {
    childNode = /*#__PURE__*/React.createElement("form", {
      onSubmit: handleSubmit,
      className: formClassName
    }, childNode);
  }

  // ---------------- scrollToFirstError ----------------
  const scrollToFirstError = React.useCallback(option => {
    if (shouldScrollToFirstErrorField && formCtx.isDirty) {
      var _descendants$values$f;
      const node = (_descendants$values$f = descendants.values().find(o => o.id && get(formCtx.errors, unmarshalNamePath(o.id)))) === null || _descendants$values$f === void 0 ? void 0 : _descendants$values$f.node;
      if (node) {
        const mergedOption = option !== null && option !== void 0 ? option : shouldScrollToFirstErrorField === true ? {
          block: 'nearest'
        } : shouldScrollToFirstErrorField;
        scrollIntoView(node, {
          scrollMode: 'if-needed',
          ...(mergedOption || {})
        });
      }
    }
  }, [descendants, formCtx.errors, formCtx.isDirty, shouldScrollToFirstErrorField]);

  /**
   * ================ Ref ================
   */
  useImperativeHandle(ref, () => Object.assign(Object.create(Object.getPrototypeOf(formCtx), Object.getOwnPropertyDescriptors(formCtx)), {
    scrollToFirstError
  }));
  return /*#__PURE__*/React.createElement(FormDescendantsProvider, {
    value: descendants
  }, /*#__PURE__*/React.createElement(ConfigProvider, {
    componentSize: sizeProp
  }, /*#__PURE__*/React.createElement(FormUIContextProvider, {
    value: {
      requiredMark,
      labelAlign,
      labelCol,
      wrapperCol,
      validateMessages,
      errorDisplay
    }
  }, /*#__PURE__*/React.createElement(FormContextProvider, {
    value: formCtx
  }, childNode))));
});
function useValidators(props) {
  const {
    label,
    rules,
    validate: validateProp
  } = props;
  const formCtxValidateMessages = useFormUIContextSelector(ctx => ctx.validateMessages);
  const locale = useLocaleReceiver('form');
  const variables = useMemo(() => {
    const _variables = {};
    if (typeof label === 'string') {
      _variables.label = label;
    } else {
      _variables.label = '';
    }
    return _variables;
  }, [label]);
  const validate = React.useMemo(() => {
    const validators = [];
    if (rules) {
      validators.push(...asArray(rules).map(rule => {
        const validator = {
          validateTrigger: rule.validateTrigger,
          validate: async (value, _name) => {
            const namePathString = marshalNamePath(_name);
            const errorMessages = await validateRule(namePathString, value, rule, {
              validateMessages: merge({}, formCtxValidateMessages, locale.defaultValidateMessage)
            }, variables);
            if (errorMessages.length > 0) {
              return errorMessages[0];
            }
          }
        };
        return validator;
      }));
    }
    if (validateProp) {
      validators.push(...asArray(validateProp));
    }
    return validators;
  }, [locale.defaultValidateMessage, formCtxValidateMessages, rules, validateProp, variables]);
  return validate;
}
function _FormItem(props) {
  const {
    noStyle = false,
    label,
    className,
    style,
    name,
    children,
    labelAlign,
    labelCol,
    rules,
    // eslint-disable-next-line no-nested-ternary
    required = rules ? Array.isArray(rules) ? rules.some(r => r.required) : rules.required : undefined,
    wrapperCol,
    help,
    validateStatus,
    tooltip,
    extra,
    hidden,
    htmlFor,
    validate: validateProp,
    labelInfoIconRenderer,
    errorDisplay,
    ...restProps
  } = props;
  const {
    error
  } = useFieldContext({
    name
  });
  const mergedValidateStatus = validateStatus !== null && validateStatus !== void 0 ? validateStatus : error ? 'error' : undefined;
  process.env.NODE_ENV !== "production" ? warning(isNonNullable(name), devWarning('FormItem', 'The `name` prop is undefined, thus this field will not be registered to the form')) : void 0;
  const validate = useValidators({
    label,
    validate: validateProp,
    rules
  });

  /**
   * Descendant
   */
  const id = isNonNullable(name) ? marshalNamePath(name) : undefined;
  const {
    register
  } = useFormDescendant(React.useMemo(() => ({
    id
  }), [id]));

  /**
   * ================ Input Content Node ================
   *
   * child type: funcChild | reactNodeChild
   *
   * noStyle: true | false
   *
   * - 如果 noStyle === false，则直接用 FormItemUI 的 ref 来 register
   * - 如果 noStyle === true，则需要判断 children 类型
   *  - 如果 children 是 ReactElement 则直接套住 FindDOMNode
   *  - 如果 children 是 function 则先交给 Field 处理，然后再套住 FindDOMNode
   */
  let inputContentNode = children;
  // 判断是否是一个 field 组件，如果没有 name 的话则认为只是一个 UI submit 而不是 field
  if (props.name !== undefined && (/*#__PURE__*/isValidElement(children) || isFunction(children))) {
    inputContentNode = /*#__PURE__*/React.createElement(Field, _extends({
      name: props.name,
      validate: validate,
      validateStatus: mergedValidateStatus
    }, restProps), (() => {
      if (!noStyle) {
        return children;
      }
      if (isFunction(children)) {
        return fieldProps => /*#__PURE__*/React.createElement(FindDOMNode, {
          domOnly: true,
          childRef: register
        }, children(fieldProps));
      }
      return /*#__PURE__*/React.createElement(FindDOMNode, _extends({}, children.props, {
        domOnly: true,
        childRef: register
      }), children);
    })());
  }
  if (noStyle) {
    return /*#__PURE__*/React.createElement(React.Fragment, null, inputContentNode);
  }
  return /*#__PURE__*/React.createElement(FormItemUI, {
    label,
    className,
    style,
    children,
    labelAlign,
    labelCol,
    required,
    wrapperCol,
    help: help !== null && help !== void 0 ? help : error,
    hidden,
    validateStatus: mergedValidateStatus,
    tooltip,
    extra,
    htmlFor,
    labelInfoIconRenderer,
    errorDisplay,
    ref: register
  }, inputContentNode);
}
export function FormList(props) {
  const {
    validate,
    rules,
    ...others
  } = props;
  const validators = useValidators({
    rules,
    validate
  });
  return /*#__PURE__*/React.createElement(MiniFormList, _extends({
    validate: validators
  }, others));
}
export { useField, useFieldContext, useForm, useFormContext, useFormContextSelector, FormSelector, FormSelectors, useFormList, FormContextProvider } from '../react-mini-form';
export const Form = BaseForm;
export const FormItem = /*#__PURE__*/memo(_FormItem, FieldPropsAreEqual);
Form.Item = FormItem;
Form.ItemUI = FormItemUI;
if (process.env.NODE_ENV !== "production") {
  FormItem.displayName = 'FormItem';
}
