// @universe-design/styles/cjs/ccm-ccm-sheet/index.light.atom-raw.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ccmtoken_sheet_minichart_professional_red_raw = exports.ccmtoken_sheet_minichart_professional_blue_raw = exports.ccmtoken_sheet_minichart_pastel_yellow_raw = exports.ccmtoken_sheet_minichart_pastel_blue_raw = exports.ccmtoken_sheet_minichart_fresh_yellow_raw = exports.ccmtoken_sheet_minichart_fresh_cyan_raw = exports.ccmtoken_sheet_minichart_bright_yellow_raw = exports.ccmtoken_sheet_minichart_bright_purple_raw = void 0;
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
exports.ccmtoken_sheet_minichart_bright_purple_raw = '97, 71, 255';
exports.ccmtoken_sheet_minichart_bright_yellow_raw = '255, 186, 10';
exports.ccmtoken_sheet_minichart_fresh_cyan_raw = '36, 196, 205';
exports.ccmtoken_sheet_minichart_fresh_yellow_raw = '255, 209, 58';
exports.ccmtoken_sheet_minichart_pastel_blue_raw = '76, 146, 173';
exports.ccmtoken_sheet_minichart_pastel_yellow_raw = '229, 193, 0';
exports.ccmtoken_sheet_minichart_professional_blue_raw = '56, 96, 174';
exports.ccmtoken_sheet_minichart_professional_red_raw = '185, 34, 30';
