// @universe-design/styles/cjs/ccm-ccm-sheet/index.light.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ccmtoken_sheet_minichart_professional_red = exports.ccmtoken_sheet_minichart_professional_blue = exports.ccmtoken_sheet_minichart_pastel_yellow = exports.ccmtoken_sheet_minichart_pastel_blue = exports.ccmtoken_sheet_minichart_fresh_yellow = exports.ccmtoken_sheet_minichart_fresh_cyan = exports.ccmtoken_sheet_minichart_bright_yellow = exports.ccmtoken_sheet_minichart_bright_purple = void 0;
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
exports.ccmtoken_sheet_minichart_bright_purple = '#6147ff';
exports.ccmtoken_sheet_minichart_bright_yellow = '#ffba0a';
exports.ccmtoken_sheet_minichart_fresh_cyan = '#24c4cd';
exports.ccmtoken_sheet_minichart_fresh_yellow = '#ffd13a';
exports.ccmtoken_sheet_minichart_pastel_blue = '#4c92ad';
exports.ccmtoken_sheet_minichart_pastel_yellow = '#e5c100';
exports.ccmtoken_sheet_minichart_professional_blue = '#3860ae';
exports.ccmtoken_sheet_minichart_professional_red = '#b9221e';
