// @universe-design/styles/cjs/ccm-ccm-sheet/index.dark.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ccmtoken_sheet_minichart_professional_red = exports.ccmtoken_sheet_minichart_professional_blue = exports.ccmtoken_sheet_minichart_pastel_yellow = exports.ccmtoken_sheet_minichart_pastel_blue = exports.ccmtoken_sheet_minichart_fresh_yellow = exports.ccmtoken_sheet_minichart_fresh_cyan = exports.ccmtoken_sheet_minichart_bright_yellow = exports.ccmtoken_sheet_minichart_bright_purple = void 0;
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
exports.ccmtoken_sheet_minichart_bright_purple = '#624bef';
exports.ccmtoken_sheet_minichart_bright_yellow = '#edb118';
exports.ccmtoken_sheet_minichart_fresh_cyan = '#1fd0da';
exports.ccmtoken_sheet_minichart_fresh_yellow = '#eebf26';
exports.ccmtoken_sheet_minichart_pastel_blue = '#4c92ad';
exports.ccmtoken_sheet_minichart_pastel_yellow = '#bea41a';
exports.ccmtoken_sheet_minichart_professional_blue = '#2b4b8a';
exports.ccmtoken_sheet_minichart_professional_red = '#972e2b';
