// @byted/mcp-client-sdk/dist/index.js
import * as __WEBPACK_EXTERNAL_MODULE__byted_service_fetch_b9e10547__ from "@byted-service/fetch";
import * as __WEBPACK_EXTERNAL_MODULE_crypto__ from "crypto";
import * as __WEBPACK_EXTERNAL_MODULE__byted_ai_paas_env_f71d18dd__ from "@byted/ai-paas-env";
import * as __WEBPACK_EXTERNAL_MODULE_joi__ from "joi";
import * as __WEBPACK_EXTERNAL_MODULE__byted_service_zti_helper_95100fdd__ from "@byted-service/zti-helper";
import * as __WEBPACK_EXTERNAL_MODULE__byted_service_logger_5ae81f0d__ from "@byted-service/logger";
const WITH_HTTP_MESH_EGRESS_UDS = 'WITH_HTTP_MESH_EGRESS_UDS';
const SERVICE_MESH_HTTP_EGRESS_ADDR = 'SERVICE_MESH_HTTP_EGRESS_ADDR';
const MESH_EGRESS_PORT = 'MESH_EGRESS_PORT';
const BYTEDMCP_DIRECT_CONNECT = 'BYTEDMCP_DIRECT_CONNECT';
function isMeshMode() {
    const directConnect = process.env[BYTEDMCP_DIRECT_CONNECT] || '';
    const withHttpMeshEgressUDS = process.env[WITH_HTTP_MESH_EGRESS_UDS] || '';
    const serviceMeshHttpEgressAddr = process.env[SERVICE_MESH_HTTP_EGRESS_ADDR] || '';
    const meshEgressPort = process.env[MESH_EGRESS_PORT] || '';
    return 'true' !== directConnect.toLowerCase() && ('true' === withHttpMeshEgressUDS.toLowerCase() && '' !== serviceMeshHttpEgressAddr || '' !== meshEgressPort);
}
const generateDefaultLogId = ()=>{
    const version = '02';
    const timestamp = Date.now().toString();
    const ipv6Placeholder = '0'.repeat(32);
    const randomPart = (0, __WEBPACK_EXTERNAL_MODULE_crypto__.randomBytes)(3).toString('hex');
    return `${version}${timestamp}${ipv6Placeholder}${randomPart}`;
};
function getDomainSuffix(region) {
    let isProdNetwork;
    isProdNetwork = 'office' === process.env.BYTED_MCP_CLIENT_SDK_NETWORK_TYPE ? false : (0, __WEBPACK_EXTERNAL_MODULE__byted_ai_paas_env_f71d18dd__.ifProdNetwork)();
    let isComplianceNetwork;
    isComplianceNetwork = 'sidecar' === process.env.COMPLIANCE_MODE;
    switch(region){
        case __WEBPACK_EXTERNAL_MODULE__byted_ai_paas_env_f71d18dd__.AI_PAAS_REGION.BOE:
            return '.mcp-boe.bytedance.net';
        case __WEBPACK_EXTERNAL_MODULE__byted_ai_paas_env_f71d18dd__.AI_PAAS_REGION.CN:
            return '.mcp.bytedance.net';
        case __WEBPACK_EXTERNAL_MODULE__byted_ai_paas_env_f71d18dd__.AI_PAAS_REGION.I18N:
            if (isProdNetwork) {
                if (isComplianceNetwork) return '.mcp.tiktok-row.org';
                return '.mcp.byteintl.net';
            }
            return '.mcp.tiktok-row.net';
        case __WEBPACK_EXTERNAL_MODULE__byted_ai_paas_env_f71d18dd__.AI_PAAS_REGION.I18NBD:
            return '.mcp-i18nbd.byteintl.net';
        case __WEBPACK_EXTERNAL_MODULE__byted_ai_paas_env_f71d18dd__.AI_PAAS_REGION.BOEI18N:
            return '.mcp-boei18n.byteintl.net';
        case __WEBPACK_EXTERNAL_MODULE__byted_ai_paas_env_f71d18dd__.AI_PAAS_REGION.EUTTP:
            return isProdNetwork ? '.mcp-eu.tiktoke.org' : '.mcp-eu.tiktok-eu.net';
        case __WEBPACK_EXTERNAL_MODULE__byted_ai_paas_env_f71d18dd__.AI_PAAS_REGION.USTTP:
            return isProdNetwork ? '.mcp-usttp.tiktokd.net' : '.mcp-usttp.tiktok-us.net';
        default:
            throw new Error(`[getDomainSuffix] Unsupported input region ${region}`);
    }
}
function getMCPHttpEndpoint({ region, serviceId }) {
    const domainSuffix = getDomainSuffix(region);
    const domainPrefix = serviceId;
    return `https://${domainPrefix}${domainSuffix}/mcp`;
}
function getNameServerEndpoint({ region, psm, env }) {
    const domainSuffix = getDomainSuffix(region);
    const domainPrefix = 'mcp';
    if (env) return `https://${domainPrefix}${domainSuffix}/v1/psm/${psm}/env/${env}`;
    return `https://${domainPrefix}${domainSuffix}/v1/psm/${psm}`;
}
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
const DEFAULT_HTTP_TIMEOUT = 300000;
const COMMON_BODY_FIEDS = {
    jsonrpc: '2.0',
    id: 1
};
class McpHttpClient {
    async initUrlWithPsm({ psm, env, region, ctx }) {
        if (region === __WEBPACK_EXTERNAL_MODULE__byted_ai_paas_env_f71d18dd__.AI_PAAS_REGION.LOCALHOST_DIRECT) {
            this.httpEndpoint = 'http://localhost:3080/mcp';
            return;
        }
        const response = await (0, __WEBPACK_EXTERNAL_MODULE__byted_service_fetch_b9e10547__["default"])(getNameServerEndpoint({
            psm,
            region,
            env
        }), {
            method: 'get',
            headers: {
                'X-Mcp-Controlplane-Request': 'true'
            },
            timeout: this.timeout,
            enableBytedtrace: false,
            logId: ctx.logId
        });
        if (!response.ok) throw new Error(`[initUrlWithPsm] HTTP Error Response: ${response.status} ${response.statusText}`);
        const serviceId = (await response.json()).data.server_id;
        const endpoint = getMCPHttpEndpoint({
            serviceId,
            region
        });
        ctx.logger.debug(`[initUrlWithPsm] Initialize httpEndpoint successfuly. psm=${psm}, endpoint=${endpoint}`);
        this.httpEndpoint = endpoint;
    }
    async listTools({ ctx }) {
        if (!this.httpEndpoint) throw new Error(`HTTP endpoint not initialized for server: ${this.serverName}`);
        const requestBody = {
            ...COMMON_BODY_FIEDS,
            method: 'tools/list',
            params: {
                _meta: {
                    headers: {},
                    params: {}
                }
            }
        };
        const requestHeaders = {
            'Content-Type': 'application/json'
        };
        for await (const beforeCallHook of this.beforeCallHooks)await beforeCallHook({
            body: requestBody,
            headers: requestHeaders
        });
        ctx.logger.debug(`[listTools] after all beforeCallHooks, requestHeaders=${JSON.stringify(requestHeaders)}, requestBody=${JSON.stringify(requestBody)}`);
        const response = await (0, __WEBPACK_EXTERNAL_MODULE__byted_service_fetch_b9e10547__["default"])(this.httpEndpoint, {
            method: 'post',
            body: JSON.stringify(requestBody),
            headers: requestHeaders,
            timeout: this.timeout,
            enableBytedtrace: false,
            logId: ctx.logId
        });
        if (!response.ok) throw new Error(`[listTools] HTTP Error Response: ${response.status} ${response.statusText}`);
        const jsonResponse = await response.json();
        ctx.logger.debug(`[listTools] List tools successfully, length=${jsonResponse.result.tools.length}, names=${jsonResponse.result.tools.map((item)=>item.name).join(',')}`);
        return jsonResponse.result.tools;
    }
    registerBeforeCallHook({ beforeCallHook, ctx }) {
        this.beforeCallHooks.push(beforeCallHook);
        ctx.logger.debug(`[registerBeforeCallHook] Added beforeCallHook for server: ${this.serverName}`);
    }
    clearBeforeCallHook({ ctx }) {
        this.beforeCallHooks = [];
        ctx.logger.debug(`[clearBeforeCallHook] Cleared beforeCallHook for server: ${this.serverName}`);
    }
    clearAfterCallHook({ ctx }) {
        this.afterCallHooks = [];
        ctx.logger.debug(`[clearAfterCallHook] Cleared afterCallHook for server: ${this.serverName}`);
    }
    close({ ctx }) {
        this.beforeCallHooks = [];
        this.afterCallHooks = [];
        ctx.logger.debug(`[close] McpHttpClient for ${this.serverName} closed.`);
    }
    registerAfterCallHook({ afterCallHook, ctx }) {
        this.afterCallHooks.push(afterCallHook);
        ctx.logger.debug(`[registerBeforeCallHook] Added afterCallHook for server: ${this.serverName}`);
    }
    async callTool(name, args) {
        if (!this.httpEndpoint) throw new Error(`HTTP endpoint not initialized for server: ${this.serverName}`);
        const requestBody = {
            ...COMMON_BODY_FIEDS,
            method: 'tools/call',
            params: {
                name,
                arguments: args,
                _meta: {
                    headers: this.headers ?? {},
                    params: this.params ?? {}
                }
            }
        };
        const requestHeaders = {
            'Content-Type': 'application/json'
        };
        for await (const beforeCallHook of this.beforeCallHooks)await beforeCallHook({
            body: requestBody,
            headers: requestHeaders
        });
        const response = await (0, __WEBPACK_EXTERNAL_MODULE__byted_service_fetch_b9e10547__["default"])(this.httpEndpoint, {
            method: 'post',
            body: JSON.stringify(requestBody),
            headers: requestHeaders,
            timeout: this.timeout,
            enableHttpMesh: this.isMeshMode,
            enableBytedtrace: false
        });
        if (!response.ok) throw new Error(`[callTool] HTTP Error Response: ${response.status} ${response.statusText}`);
        const jsonResponse = await response.json();
        if (!(null == jsonResponse ? void 0 : jsonResponse.result)) throw new Error(`[callTool] Got an empty jsonResponse.result: jsonResponse=${JSON.stringify(jsonResponse)}`);
        for await (const afterCallHook of this.afterCallHooks)await afterCallHook({
            body: jsonResponse
        });
        return jsonResponse.result;
    }
    constructor({ serverName, connection }){
        _define_property(this, "serverName", void 0);
        _define_property(this, "timeout", void 0);
        _define_property(this, "httpEndpoint", void 0);
        _define_property(this, "psm", void 0);
        _define_property(this, "isMeshMode", void 0);
        _define_property(this, "headers", void 0);
        _define_property(this, "params", void 0);
        _define_property(this, "beforeCallHooks", []);
        _define_property(this, "afterCallHooks", []);
        this.serverName = serverName;
        this.timeout = connection.timeout || DEFAULT_HTTP_TIMEOUT;
        this.headers = connection.headers;
        this.params = connection.params;
        if (connection.url) this.httpEndpoint = connection.url;
        else this.psm = connection.psm;
        this.isMeshMode = isMeshMode();
    }
}
var package_namespaceObject = JSON.parse('{"u2":"@byted/mcp-client-sdk","i8":"1.3.6"}');
function mcpClient_define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class ServerInfo {
    constructor(client){
        mcpClient_define_property(this, "tools", void 0);
        mcpClient_define_property(this, "client", void 0);
        this.client = client;
        this.tools = [];
    }
}
class BytedMcpClient {
    validateConnections(connections) {
        const connectionSchema = __WEBPACK_EXTERNAL_MODULE_joi__["default"].object({
            transport: __WEBPACK_EXTERNAL_MODULE_joi__["default"].string().valid('http').required(),
            url: __WEBPACK_EXTERNAL_MODULE_joi__["default"].string().uri().pattern(/^https?:\/\//),
            psm: __WEBPACK_EXTERNAL_MODULE_joi__["default"].string().pattern(/^[\w]+\.[\w]+\.[\w]+$/),
            env: __WEBPACK_EXTERNAL_MODULE_joi__["default"].string(),
            headers: __WEBPACK_EXTERNAL_MODULE_joi__["default"].object(),
            params: __WEBPACK_EXTERNAL_MODULE_joi__["default"].object(),
            timeout: __WEBPACK_EXTERNAL_MODULE_joi__["default"].number(),
            mcpGatewayRegion: __WEBPACK_EXTERNAL_MODULE_joi__["default"].when('psm', {
                is: __WEBPACK_EXTERNAL_MODULE_joi__["default"].exist(),
                then: __WEBPACK_EXTERNAL_MODULE_joi__["default"].string().valid(...Object.values(__WEBPACK_EXTERNAL_MODULE__byted_ai_paas_env_f71d18dd__.AI_PAAS_REGION)).required().messages({
                    'any.required': '`mcpGatewayRegion` is required when using PSM connection',
                    'any.only': '`mcpGatewayRegion` must be one of BOE, CN, or I18N'
                }),
                otherwise: __WEBPACK_EXTERNAL_MODULE_joi__["default"].string().valid(...Object.values(__WEBPACK_EXTERNAL_MODULE__byted_ai_paas_env_f71d18dd__.AI_PAAS_REGION)).optional()
            })
        }).xor('url', 'psm').messages({
            'object.xor': 'Must specify exactly one of `url` or `psm`',
            'string.pattern.base': 'Invalid PSM format'
        });
        if (!connections || 0 === Object.keys(connections).length) throw new Error('Connections configuration cannot be empty');
        for (const [name, conn] of Object.entries(connections)){
            const { error } = connectionSchema.validate(conn, {
                abortEarly: false
            });
            if (error) throw new Error(`[validateConnections] Configuration error for connection ${name}: ${error.details.map((d)=>d.message).join('; ')}`);
        }
    }
    async connectServer({ serverName, connection, ctx }) {
        const client = new McpHttpClient({
            serverName,
            connection
        });
        if (connection.psm) try {
            await client.initUrlWithPsm({
                psm: connection.psm,
                env: connection.env,
                region: connection.mcpGatewayRegion ?? __WEBPACK_EXTERNAL_MODULE__byted_ai_paas_env_f71d18dd__.AI_PAAS_REGION.CN,
                ctx
            });
        } catch (e) {
            const { message } = e;
            throw new Error(`Failed to init PSM for ${serverName}: ${message}`);
        }
        const serverInfo = new ServerInfo(client);
        this.servers.set(serverName, serverInfo);
    }
    listLoadedTools({ ctx }) {
        const tools = Array.from(this.servers.values()).flatMap((server)=>server.tools);
        ctx.logger.debug(`[listLoadedTools] Load tools successfully, length=${tools.length}, names=${tools.map((item)=>item.name).join(',')}`);
        return tools;
    }
    async callTool(toolName, args) {
        var _tool_modelConfig;
        const tool = this.findTool(toolName);
        if (!tool) throw new Error(`Tool ${toolName} not found`);
        const serverName = null == (_tool_modelConfig = tool.modelConfig) ? void 0 : _tool_modelConfig._mcpServerName;
        if ('string' != typeof serverName) throw new Error(`[callTool] invalid serverName: ${serverName}`);
        const server = this.servers.get(serverName);
        if (!server) throw new Error(`[callTool] Server ${serverName} not found for tool ${toolName}`);
        return await server.client.callTool(toolName, args);
    }
    registerBeforeCallHook({ serverName, beforeCallHook, ctx }) {
        if ('*' !== serverName) {
            const server = this.servers.get(serverName);
            if (!server) throw new Error(`[registerBeforeCallHook] Server ${serverName} not found.`);
            server.client.registerBeforeCallHook({
                beforeCallHook,
                ctx
            });
        } else for (const [, serverInfo] of this.servers)serverInfo.client.registerBeforeCallHook({
            beforeCallHook,
            ctx
        });
    }
    clearBeforeCallHook({ serverName, ctx }) {
        if ('*' !== serverName) {
            const server = this.servers.get(serverName);
            if (!server) throw new Error(`[registerBeforeCallHook] Server ${serverName} not found.`);
            server.client.clearBeforeCallHook({
                ctx
            });
        } else for (const [, serverInfo] of this.servers)serverInfo.client.clearBeforeCallHook({
            ctx
        });
    }
    registerAfterCallHook({ serverName, afterCallHook, ctx }) {
        if ('*' !== serverName) {
            const server = this.servers.get(serverName);
            if (!server) throw new Error(`[registerAfterCallHook] Server ${serverName} not found.`);
            server.client.registerAfterCallHook({
                ctx,
                afterCallHook
            });
        } else for (const [, serverInfo] of this.servers)serverInfo.client.registerAfterCallHook({
            ctx,
            afterCallHook
        });
    }
    findTool(toolName) {
        const serverArray = Array.from(this.servers.values());
        for (const server of serverArray){
            const found = server.tools.find((t)=>t.name === toolName);
            if (found) return found;
        }
    }
    async connectAll({ ctx }) {
        ctx.logger.debug(`[connectAll] ${package_namespaceObject.u2}@${package_namespaceObject.i8} is attempting to connect to all MCP servers`);
        await Promise.all(Object.entries(this.connections).map(([name, conn])=>this.connectServer({
                serverName: name,
                connection: conn,
                ctx
            })));
        ctx.logger.debug(`[connectAll] ${package_namespaceObject.u2}@${package_namespaceObject.i8} has connected to all MCP servers`);
    }
    close({ ctx }) {
        for (const [serverName, serverInfo] of this.servers){
            serverInfo.client.close({
                ctx
            });
            ctx.logger.debug(`[close] Closed connection to server: ${serverName}`);
        }
        this.servers.clear();
        ctx.logger.debug('[close] All MCP client connections closed.');
    }
    constructor(connections){
        mcpClient_define_property(this, "servers", void 0);
        mcpClient_define_property(this, "connections", void 0);
        mcpClient_define_property(this, "loadAllTools", async ({ ctx })=>{
            if (0 === this.servers.size) throw new Error("[loadAllTools] No server is connected");
            for (const [serverName, serverInfo] of this.servers){
                ctx.logger.debug(`[loadAllTools] loading ${serverName}'s tool`);
                const tools = await serverInfo.client.listTools({
                    ctx
                });
                serverInfo.tools = tools.map((tool)=>({
                        ...tool,
                        modelConfig: {
                            ...tool.modelConfig,
                            _mcpServerName: serverName
                        }
                    }));
            }
            return this.listLoadedTools({
                ctx
            });
        });
        this.validateConnections(connections);
        this.connections = connections;
        this.servers = new Map();
    }
}
function logger_define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class BaseLogger {
}
const bytedLogger = new __WEBPACK_EXTERNAL_MODULE__byted_service_logger_5ae81f0d__["default"]();
class BytedServiceLogger extends BaseLogger {
    log(level, input) {
        bytedLogger.log({
            message: input,
            logId: this.logId,
            level
        });
    }
    debug(input) {
        this.log('debug', input);
    }
    info(input) {
        this.log('info', input);
    }
    warn(input) {
        this.log('warn', input);
    }
    error(input) {
        this.log('error', input);
    }
    trace(input) {
        this.log('trace', input);
    }
    constructor({ logId }){
        super(), logger_define_property(this, "logId", void 0);
        this.logId = logId;
    }
}
class ConsoleErrorLogger extends BaseLogger {
    debug(input) {
        console.error(input);
    }
    info(input) {
        console.error(input);
    }
    warn(input) {
        console.error(input);
    }
    error(input) {
        console.error(input);
    }
    trace(input) {
        console.error(input);
    }
}
const generateBytedServiceLoggerWithLogId = ({ logId })=>new BytedServiceLogger({
        logId
    });
const generateConsoleErrorLogger = ()=>new ConsoleErrorLogger();
const ztiHelper = new __WEBPACK_EXTERNAL_MODULE__byted_service_zti_helper_95100fdd__.ZtiJWTHelper({
    logger: generateConsoleErrorLogger()
});
const addZtiTokenHook = async ({ headers })=>{
    try {
        headers['x-zti-token'] = await ztiHelper.getJwtSVID();
    } catch (error) {
        console.error(`[addZtiTokenHook] getJwtSVID failed, ${error}`);
    }
};
const generateAddHttpHeadersHook = ({ httpHeaders })=>{
    const hook = async ({ headers })=>{
        const newHeaders = {
            ...headers,
            ...httpHeaders
        };
        Object.assign(headers, newHeaders);
    };
    return hook;
};
const generateAddCallToolHeadersHook = ({ callToolHeaders, toolName })=>{
    const hook = async ({ body })=>{
        if ('tools/call' === body.method && (body.params.name === toolName || '*' === toolName)) body.params._meta.headers = {
            ...body.params._meta.headers,
            ...callToolHeaders
        };
    };
    return hook;
};
const generateAddCallToolParamsHook = ({ callToolParams, toolName })=>{
    const hook = async ({ body })=>{
        if ('tools/call' === body.method && body.params.name === toolName) body.params._meta.params = {
            ...body.params._meta.params,
            ...callToolParams
        };
    };
    return hook;
};
const generatePrintHeaderAndBodyBeforeCallHook = ({ ctx })=>{
    const hook = async ({ body, headers })=>{
        ctx.logger.info(`[beforeCallHook-print]: headers=${JSON.stringify(headers)}, body=${JSON.stringify(body)}`);
    };
    return hook;
};
const generatePrintBodyAfterCallHook = ({ ctx })=>{
    const hook = async ({ body })=>{
        ctx.logger.info(`[afterCallHook-print]: body=${JSON.stringify(body)}`);
    };
    return hook;
};
export { BytedMcpClient, addZtiTokenHook, generateAddCallToolHeadersHook, generateAddCallToolParamsHook, generateAddHttpHeadersHook, generateBytedServiceLoggerWithLogId, generateConsoleErrorLogger, generateDefaultLogId, generatePrintBodyAfterCallHook, generatePrintHeaderAndBodyBeforeCallHook };
