// @byted-service/consul/lib/consul.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Consul = void 0;
const fs_1 = require("fs");
const net_1 = require("net");
const env_1 = require("@byted-service/env");
const debug_1 = require("./debug");
const errors_1 = require("./errors");
const helper_1 = require("./helper");
const consts_1 = require("./consts");
const bytedSdUdsPathExists = (0, fs_1.existsSync)(consts_1.bytedSdUdsPath);
const consulUnixSockExists = (0, fs_1.existsSync)(consts_1.consulUnixSock);
class Consul {
    constructor(options = {}) {
        this.cache = {};
        if (options.socketPath) {
            this.address = options.socketPath;
        }
        else {
            const host = options.host;
            const port = Number(options.port) || consts_1.defaultPort;
            if (host) {
                this.address = (0, net_1.isIPv6)(host)
                    ? `[${host}]:${port}`
                    : `${host}:${port}`;
            }
            else if (bytedSdUdsPathExists) {
                this.address = consts_1.bytedSdUdsPath;
            }
            else if (consts_1.defaultHost) {
                this.address = (0, net_1.isIPv6)(consts_1.defaultHost)
                    ? `[${consts_1.defaultHost}]:${port}`
                    : `${consts_1.defaultHost}:${port}`;
            }
            else if (consulUnixSockExists) {
                this.address = consts_1.consulUnixSock;
            }
            else {
                this.address = '127.0.0.1:2280';
            }
        }
        this.cacheTime = options.cacheTime || consts_1.defaultCacheTime;
        this.logger = options.logger || consts_1.defaultLogger;
        this.enableWarning =
            typeof options.enableWarning === 'boolean'
                ? options.enableWarning
                : (0, env_1.isInTCE)();
        this.request = (0, helper_1.createRequest)(this.address);
        if (options.debug) {
            (0, debug_1.wrapDebug)(this, options);
        }
    }
    async lookup(psm, options = {}) {
        const cluster = options.cluster === '*' ? '' : options.cluster; // 兼容性处理 cluster === *
        const idc = options.idc === '-' ? '' : options.idc; // 兼容性处理 idc === -
        const addrfam = options.addrfam;
        const unique = options.unique;
        const limit = options.limit;
        const nocache = options.nocache;
        let consulHash = options.consulHash;
        const timeout = options.timeout;
        const useCache = !(limit && limit > 0); // limit 底层会使用抽牌算法，因此带 limit 的查询结果无法缓存
        const key = `${psm}|${cluster}|${idc}|${addrfam}|${unique}`;
        let item = useCache ? this.cache[key] : undefined;
        if (item) {
            if (Date.now() < item.updatedAt + this.cacheTime && !nocache) {
                return item.endpoints;
            }
            if (item.consulHash) {
                consulHash = consulHash || item.consulHash;
            }
        }
        let url = `/v1/lookup/name?name=${idc && !psm.includes('.service.') ? `${psm}.service.${idc}` : psm}`;
        if (limit && limit > 0) {
            url += `&limit=${limit}`;
        }
        if (cluster) {
            url += `&cluster=${cluster}`;
        }
        if (addrfam) {
            url += `&addr-family=${addrfam}`;
        }
        if (unique && addrfam === 'dual-stack') {
            url += `&unique=${unique}`;
        }
        const requestOptions = { timeout };
        if (consulHash) {
            requestOptions.headers = requestOptions.headers || {};
            requestOptions.headers[consts_1.consulHashHeader] = consulHash;
        }
        try {
            const resp = await this.request(url, requestOptions);
            if (resp.statusCode !== 200) {
                if (resp.statusCode === 304) {
                    if (!item) {
                        throw new errors_1.ConsulError('Consul agent returns 304, but cache not exists, there is some bug. PSM: %s, options: %j, address: %s', psm, options, this.address);
                    }
                    item.updatedAt = Date.now();
                    return item.endpoints;
                }
                throw new errors_1.ConsulError('Failed to get result from consul agent. PSM: %s, options: %j, address: %s, statusCode: %d, statusMessage: %s', psm, options, this.address, resp.statusCode, resp.body);
            }
            const body = JSON.parse(resp.body);
            let endpoints = [];
            for (let i = 0; i < body.length; i++) {
                const { Host, Port, Tags = {} } = body[i];
                const address = (0, net_1.isIPv6)(Host)
                    ? `[${Host}]:${Port}`
                    : `${Host}:${Port}`;
                const weight = Number(Tags.weight);
                endpoints.push({
                    host: Host,
                    port: Port,
                    uri: address,
                    addr: address,
                    cluster: Tags.cluster || '',
                    env: Tags.env || '',
                    weight: Number.isNaN(weight) ? 50 : weight,
                    tags: Tags,
                });
            }
            if (cluster) {
                const eps = [];
                for (let i = 0; i < endpoints.length; i++) {
                    if (endpoints[i].cluster === cluster) {
                        eps.push(endpoints[i]);
                    }
                }
                endpoints = eps;
            }
            item = {
                updatedAt: Date.now(),
                endpoints,
                consulHash: resp.headers[consts_1.consulHashHeader] || '',
            };
            if (useCache) {
                this.cache[key] = item;
            }
            return item.endpoints;
        }
        catch (err) {
            if (item && !nocache) {
                item.updatedAt = Date.now();
                if (this.enableWarning) {
                    this.logger.warn('[consul] Failed to get result from consul agent, use local cache. PSM: %s, options: %j, address: %s, cache: %j. %s', psm, options, this.address, item, err);
                }
                return item.endpoints;
            }
            if (err.message.includes('ETIMEOUT') ||
                err.message.includes('ECONNREFUSED')) {
                throw new errors_1.ConsulError('Failed to get result from consul agent, this may be caused by the wrong address. ' +
                    'Please set CONSUL_HTTP_HOST or ROUTE_IP in development environment, see https://code.byted.org/nodejs/byted-service/blob/master/packages/consul/README.md#faq for details. ' +
                    'PSM: %s, options: %j, address: %s, environments: %j, error: %s', psm, options, this.address, {
                    CONSUL_HTTP_HOST: process.env.CONSUL_HTTP_HOST || '',
                    MY_HOST_IP: process.env.MY_HOST_IP || '',
                    TCE_HOST_IP: process.env.TCE_HOST_IP || '',
                    MY_HOST_IPV6: process.env.MY_HOST_IPV6 || '',
                    ROUTE_IP: process.env.ROUTE_IP ||
                        /* istanbul ignore next */ '',
                }, String(err));
            }
            else if (err instanceof errors_1.ConsulError) {
                throw err;
            }
            else {
                throw new errors_1.ConsulError('Failed to get result from consul agent. PSM: %s, options: %j, address: %s, error: %s', psm, options, this.address, String(err));
            }
        }
    }
}
exports.Consul = Consul;
