// @universe-design/react/src/collapse/CollapsePanel.tsx
import type { ReactNode } from 'react';
import React from 'react';
import clsx from 'clsx';
import noop from 'lodash/noop';
import stylex from '@universe-design/stylex';
import { PanelContent } from './PanelContent';
import DownBoldOutlined from '@universe-design/icons-react/DownBoldOutlined';
import ExpandRightFilled from '@universe-design/icons-react/ExpandRightFilled';
import { CSSTransition } from 'react-transition-group';
import { collapseTransitionProps } from './transtion';
import { IconButton } from '../button';
import { cssTypographyHeadline } from '../stylex';

import './style';
/**
 *
 * @public
 */
export interface PanelProps {
  /**
   * 面板是否展开
   * @defaultValue false
   */
  isActive?: boolean;
  /**
   * 对应`activeKey`
   */
  key: React.Key;
  /**
   * 是否可折叠或指定可折叠触发区域
   */
  collapsible?: 'header' | 'disabled';
  /**
   * 设置图标位置
   * @defaultValue `right`
   */
  expandIconPosition?: 'left' | 'right';
  /**
   * 自定义类名前缀
   */
  prefixCls?: string;
}

function defaultExpandIcon(props: PanelProps, customIcon?: ReactNode) {
  const { expandIconPosition, isActive, prefixCls, collapsible } = props;
  let icon = null;
  let size;
  if (expandIconPosition === 'left') {
    icon = customIcon || <ExpandRightFilled style={{ transform: isActive ? 'rotate(90deg)' : '' }} />;
    size = 10;
  } else {
    icon = customIcon || <DownBoldOutlined style={{ transform: isActive ? 'rotate(180deg)' : '' }} />;
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    size = 16;
  }
  return (
    <IconButton
      className={clsx(`${prefixCls}-expand-icon`, {
        [`${prefixCls}-expand-icon--noaction`]: collapsible === 'header',
        [`${prefixCls}-expand-icon-unfold`]: isActive,
      })}
      icon={icon}
      size={size}
    />
  );
}

/**
 * Collapse.Panel 的属性
 * @public
 */
export interface CollapsePanelProps extends React.PropsWithChildren<unknown> {
  /**
   * 节点样式
   */
  style?: React.CSSProperties;
  /**
   * 节点类名
   */
  className?: string;
  /**
   * 是否可折叠或指定可折叠触发区域
   */
  collapsible?: 'header' | 'disabled';
  /**
   * 自定义渲染每个面板右上角的内容
   */
  extra?: React.ReactNode;
  /**
   * 被隐藏时是否渲染 DOM 结构
   * @defaultValue false
   */
  forceRender?: boolean;
  header: React.ReactNode;
  /**
   * 对应`activeKey`
   */
  key: React.Key;
  /**
   * 是否展示当前面板上的箭头
   */
  showArrow?: boolean;
  /**
   * 是否是激活状态
   */
  isActive?: boolean;
  /**
   * 点击事件
   */
  onItemClick?: (key: string) => void;
  panelKey?: React.Key;
  expandIcon?: (props: PanelProps) => React.ReactNode;
  destroyInactivePanel?: boolean;
  expandIconPosition?: 'left' | 'right';
  prefixCls?: string;
  accordion?: boolean;
}

export const CollapsePanel: React.FC<CollapsePanelProps> = props => {
  const {
    collapsible,
    children,
    isActive,
    panelKey,
    onItemClick = noop,
    header,
    extra,
    expandIcon,
    showArrow = true,
    forceRender,
    destroyInactivePanel,
    expandIconPosition,
    prefixCls,
    style,
    className,
    accordion,
  } = props;
  const disabled = collapsible === 'disabled';
  const actualActive = !disabled && isActive;

  const renderExpandIcon = () => {
    if (showArrow) {
      const panelProps = {
        isActive: actualActive,
        collapsible,
        key: panelKey as React.Key,
        expandIconPosition,
        prefixCls,
      };
      const icon = defaultExpandIcon(panelProps, expandIcon && expandIcon(panelProps));

      if (!React.isValidElement(icon)) {
        return icon;
      }

      return React.cloneElement(icon as React.ReactElement, {
        className: clsx((icon as React.ReactElement).props.className, `${prefixCls}-arrow`),
      });
    }
    return null;
  };

  const handleClick = () => {
    if (disabled) {
      return;
    }
    onItemClick(String(panelKey));
  };

  return (
    <div
      style={style}
      className={clsx(
        `${prefixCls}-item`,
        {
          [`${prefixCls}-item-disabled`]: disabled,
          [`${prefixCls}-item-no-arrow`]: !showArrow,
          [`${prefixCls}-item-active`]: actualActive,
        },
        className,
      )}
    >
      <div
        className={`${prefixCls}-header`}
        onClick={() => {
          collapsible !== 'header' && handleClick();
        }}
        role={accordion ? 'tab' : 'button'}
        aria-expanded={actualActive}
      >
        {renderExpandIcon()}
        <span
          className={clsx(`${prefixCls}-header-text`, stylex(cssTypographyHeadline))}
          onClick={() => collapsible === 'header' && handleClick()}
        >
          {header}
        </span>
        {extra && <div className={`${prefixCls}-extra`}>{extra}</div>}
      </div>
      <CSSTransition
        {...collapseTransitionProps}
        in={actualActive}
        mountOnEnter={!forceRender}
        unmountOnExit={destroyInactivePanel && !forceRender}
      >
        {(state: string) => (
          <PanelContent
            isActive={actualActive}
            forceRender={forceRender}
            prefixCls={prefixCls}
            className={clsx({
              [`${prefixCls}-content-hidden`]: state === 'exited' && !actualActive,
              [`${prefixCls}-content-transition`]: state === 'entering' || state === 'exiting',
            })}
            role={accordion ? 'tabpanel' : undefined}
          >
            {children}
          </PanelContent>
        )}
      </CSSTransition>
    </div>
  );
};
