// @universe-design/react/src/collapse/Collapse.tsx
import React, { isValidElement, useEffect, useState } from 'react';
import clsx from 'clsx';
import { useCssScope } from '../config-provider/hooks';
import type { CollapsePanelProps, PanelProps } from './CollapsePanel';
import { toFlatArray } from '../util';
import noop from 'lodash/noop';

import './style';

function getActiveKeysArray(activeKey?: React.Key | React.Key[]): React.Key[] {
  if (!activeKey) {
    return [];
  }
  let currentActiveKey = activeKey;
  if (!Array.isArray(currentActiveKey)) {
    currentActiveKey = currentActiveKey ? [currentActiveKey] : [];
  }
  return currentActiveKey.map(key => String(key));
}

/**
 * @public
 */
export interface CollapseProps extends React.PropsWithChildren<unknown> {
  /**
   * 节点样式
   */
  style?: React.CSSProperties;
  /**
   * 节点类名
   */
  className?: string;
  /**
   * 是否设置手风琴模式
   * @defaultValue false
   */
  accordion?: boolean;
  /**
   * 当前激活 tab 面板的`key`
   */
  activeKey?: React.Key | React.Key[];
  /**
   * 初始化选中面板的`key`
   */
  defaultActiveKey?: React.Key | React.Key[];
  /**
   * 销毁折叠隐藏的面板
   * @defaultValue false
   */
  destroyInactivePanel?: boolean;
  /**
   * 自定义切换图标
   */
  expandIcon?: (panelProps: PanelProps) => React.ReactNode;
  /**
   * 设置图标位置
   * @defaultValue `right`
   */
  expandIconPosition?: 'left' | 'right';
  /**
   * 切换面板的回调
   */
  onChange?: (key: string | string[]) => void;
  /**
   * 是否可折叠或指定可折叠触发区域
   */
  collapsible?: 'header' | 'disabled';
  /**
   * 自定义类名前缀
   */
  prefixCls?: string;
  /**
   * 是否突出 header
   */
  type?: 'default' | 'obvious';
}

export const Collapse: React.FC<CollapseProps> = props => {
  const {
    prefixCls: prefixClsInProps,
    expandIconPosition = 'right',
    className,
    children,
    collapsible,
    accordion,
    activeKey: activeKeyProps,
    defaultActiveKey,
    destroyInactivePanel,
    onChange = noop,
    expandIcon,
    style,
    type = 'default',
  } = props;
  const [activeKey, setActiveKey] = useState<React.Key[]>(getActiveKeysArray(activeKeyProps ?? defaultActiveKey));
  const cssScope = useCssScope('collapse');
  const prefixCls = prefixClsInProps || cssScope();

  const collapseClassName = clsx(prefixCls, [`${prefixCls}-icon-position-${expandIconPosition}`], className, {
    [`${prefixCls}-obvious-title`]: type === 'obvious',
  });

  useEffect(() => {
    if (activeKeyProps) {
      setActiveKey(getActiveKeysArray(activeKeyProps));
    }
  }, [activeKeyProps]);

  const updateActiveKey = (newActiveKey: React.Key[]) => {
    if (!('activeKey' in props)) {
      setActiveKey(newActiveKey);
    }
    onChange(accordion ? newActiveKey[0] : newActiveKey);
  };

  const handleItemClick = (panelKey: string) => {
    let newActiveKey: React.Key[] = [];
    if (accordion) {
      newActiveKey = activeKey[0] === panelKey ? [] : [panelKey];
    } else {
      const isActive = activeKey.indexOf(panelKey) > -1;
      if (isActive) {
        newActiveKey = activeKey.filter(key => key !== panelKey);
      } else {
        newActiveKey = activeKey.concat(panelKey);
      }
    }

    updateActiveKey(newActiveKey);
  };

  const getItems = () =>
    toFlatArray(children).map((child, index) => {
      if (!isValidElement(child) || typeof child.type === 'string') {
        return child;
      }

      const { collapsible: childCollapsible, destroyInactivePanel: childDestroyInactivePanel } = child.props;
      const mergeCollapsible = childCollapsible ?? collapsible;
      const mergeDestroyInactivePanel = childDestroyInactivePanel ?? destroyInactivePanel;
      const key = child.key || String(index);
      let isActive = false;
      if (accordion) {
        isActive = activeKey[0] === key;
      } else {
        isActive = activeKey.indexOf(key) > -1;
      }
      const childProps: CollapsePanelProps = {
        ...child.props,
        key,
        panelKey: key,
        collapsible: mergeCollapsible,
        isActive,
        onItemClick: handleItemClick,
        destroyInactivePanel: mergeDestroyInactivePanel,
        expandIconPosition,
        prefixCls,
        expandIcon,
        accordion,
      };
      return React.cloneElement(child, childProps);
    });

  return (
    <div style={style} className={collapseClassName} role={accordion ? 'tablist' : undefined}>
      {getItems()}
    </div>
  );
};
