// @universe-design/styles/esm/vc-livestream/index.dark.atom.js
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
export const vctoken_livestream_bg_addnew_pressed = '#505050';
export const vctoken_livestream_bg_addscene_hover = 'rgba(235, 235, 235, 0.15)';
export const vctoken_livestream_bg_addscene_pressed = 'rgba(235, 235, 235, 0.2)';
export const vctoken_livestream_bg_applaud_emphasis = '#434343';
export const vctoken_livestream_bg_btn = '#373737';
export const vctoken_livestream_bg_btn_pressed = 'rgba(235, 235, 235, 0.2)';
export const vctoken_livestream_bg_btn_selected = 'rgba(235, 235, 235, 0.15)';
export const vctoken_livestream_bg_cam_off = '#505050';
export const vctoken_livestream_bg_comment_fg_hover = 'rgba(54, 56, 60, 0)';
export const vctoken_livestream_bg_comment_hover = 'rgba(55, 57, 60, 0.98)';
export const vctoken_livestream_bg_content = '#1f1f1f';
export const vctoken_livestream_bg_core = '#0a0a0a';
export const vctoken_livestream_bg_core_speaker = '#1a1a1a';
export const vctoken_livestream_bg_gradient_fg_emphasis = 'rgba(0, 0, 0, 0)';
export const vctoken_livestream_bg_input_activity = '#292929';
export const vctoken_livestream_bg_input_emphasis = 'rgba(255, 255, 255, 0.05)';
export const vctoken_livestream_bg_interaction = '#252525';
export const vctoken_livestream_bg_members = '#191919';
export const vctoken_livestream_bg_player_button_emphasis = 'rgba(31, 35, 41, 0.8)';
export const vctoken_livestream_bg_player_fill_emphasis = '#000000';
export const vctoken_livestream_bg_reaction_left = '#3370ff';
export const vctoken_livestream_bg_reaction_right = '#7edafb';
export const vctoken_livestream_bg_speaker = 'rgba(25, 66, 148, 0.7)';
export const vctoken_livestream_bg_tag = '#3370eb';
export const vctoken_livestream_bg_tag_livestreaming = '#572120';
export const vctoken_livestream_bg_tag_testing = '#2c3171';
export const vctoken_livestream_bg_theme_emphasis = '#27292d';
export const vctoken_livestream_bg_top_set = 'rgba(255, 255, 255, 0.1)';
export const vctoken_livestream_bg_top_set_1 = '#292929';
export const vctoken_livestream_bg_top_set_2 = '#36383c';
export const vctoken_livestream_fill_btn1 = '#505050';
export const vctoken_livestream_fill_btn2 = '#5f5f5f';
export const vctoken_livestream_fill_min_player_mask = 'rgba(0, 0, 0, 0.45)';
export const vctoken_livestream_fill_new_tips_hover_web = '#292929';
export const vctoken_livestream_fill_new_tips_normal_web = '#373737';
export const vctoken_livestream_fill_playbar_volume_tip_hover = 'rgba(76, 136, 255, 0.4)';
export const vctoken_livestream_fill_player_button_hover_1 = 'rgba(76, 136, 255, 0.2)';
export const vctoken_livestream_fill_player_button_hover_2 = 'rgba(235, 235, 235, 0.1)';
export const vctoken_livestream_fill_player_button_pressed_1 = 'rgba(76, 136, 255, 0.3)';
export const vctoken_livestream_fill_player_button_pressed_2 = 'rgba(235, 235, 235, 0.2)';
export const vctoken_livestream_fill_player_panel_activity = 'rgba(31, 35, 41, 0.9)';
export const vctoken_livestream_fill_player_tag_finish = '#a15317';
export const vctoken_livestream_fill_player_tag_live = '#f05b56';
export const vctoken_livestream_fill_preview_color_1 = '#c3c5ca';
export const vctoken_livestream_fill_preview_color_2 = '#979ca2';
export const vctoken_livestream_fill_tab_color_1 = 'rgba(31, 31, 31, 0.97)';
export const vctoken_livestream_fill_tab_color_2 = 'rgba(31, 31, 31, 0)';
export const vctoken_livestream_fill_top_set_border = 'rgba(255, 255, 255, 0.06)';
export const vctoken_livestream_icon_btn_selected = '#e8e8e8';
export const vctoken_livestream_icon_btn_unselected = '#a6a6a6';
export const vctoken_livestream_icon_more = '#5f5f5f';
export const vctoken_livestream_icon_more_hover = 'rgba(235, 235, 235, 0.25)';
export const vctoken_livestream_icon_pressed = '#e8e8e8';
export const vctoken_livestream_line_btn_divider = 'rgba(235, 235, 235, 0.2)';
export const vctoken_livestream_line_btn_selected = 'rgba(235, 235, 235, 0.2)';
export const vctoken_livestream_line_mainscreen = 'rgba(235, 235, 235, 0.15)';
export const vctoken_livestream_line_speaker = '#4c88ff';
export const vctoken_livestream_player_shadow_emphasis = '0px 0px 1px rgba(0, 0, 0, 0.5)';
export const vctoken_livestream_player_shadow_emphasis_1 = '0px 4px 8px rgba(31, 35, 41, 0.06), 0px 8px 24px 8px rgba(31, 35, 41, 0.04), 0px 6px 12px rgba(31, 35, 41, 0.04)';
export const vctoken_livestream_player_shadow_emphasis_2 = '0px 1px 4px rgba(0, 0, 0, 0.2)';
export const vctoken_livestream_player_text_shadow_1 = '0px 0px 1px rgba(0, 0, 0, 0.3)';
export const vctoken_livestream_preview_inner_shadow_default = '0px 4px 24px rgba(187, 191, 196, 0.18)';
export const vctoken_livestream_preview_shadow_ash_blue = '0px 8px 30px rgba(14, 24, 40, 0.26)';
export const vctoken_livestream_preview_shadow_default = '0px 8px 48px rgba(31, 35, 41, 0.12)';
export const vctoken_livestream_text_btn_selected = '#ebebeb';
export const vctoken_livestream_text_btn_unselected = '#a6a6a6';
export const vctoken_livestream_text_logo_audience = '#ffffff';
export const vctoken_livestream_text_playbar_selected = '#4c88ff';
export const vctoken_livestream_text_pressed = '#ebebeb';
export const vctoken_livestream_text_reaction_number = '#6425d0';
export const vctoken_livestream_text_tag = '#ffffff';
export const vctoken_livestream_web_commemt_inset_shadow = 'inset -1px 1px rgba(255, 255, 255, 0.08)';
export const vctoken_livestream_web_title_inset_shadow = 'inset 0px 1px rgba(255, 255, 255, 0.08)';
