// @universe-design/react/src/tabs/ScrollableTabsHeader.tsx
/* eslint-disable react/no-unknown-property */
/** @jsxRuntime classic */
/** @jsx jsx */
/** @jsxFrag React.Fragment */
import stylex from '@universe-design/stylex';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx } from '@universe-design/stylex/react';
import React from 'react';
import { Tab } from './Tab';
import { TabProvider } from './TabContext';
import { useTabsContext } from './TabsContext';
import LeftOutlined from '@universe-design/icons-react/LeftOutlined';
import RightOutlined from '@universe-design/icons-react/RightOutlined';
import { TabsInkBar } from './TabInk';
import { composeRefs, useResizeObserver } from '@universe-design/react-utilities';
import { forwardHTMLRef } from '../util/forwardRef';
import { IconButton } from '../button';
import { resolveSizingBehavior } from './sizing_behavior';

const defaultScrollOption: ScrollIntoViewOptions = {
  // 外层有滚动容器的话，外层不跟随滚动
  block: 'nearest',
  behavior: 'smooth',
};

export const ScrollableTabsHeader = forwardHTMLRef<
  {
    /**
     * 标签下划线的偏移量
     */
    inkOffsetY: number;
  },
  'div'
>(({ inkOffsetY }, ref) => {
  const { rootCssScope, panes, scrollButtonGutter, tabElementMapRef, type, rootTabsSizingBehavior } = useTabsContext();
  const [scrollButtonVisible, setScrollButtonVisible] = React.useState(false);
  const [scrollButtonDisabled, setScrollButtonDisabled] = React.useState([true, false]);

  const scrollMaskWidth = type === 'line' ? 22 : 0;

  const resizeHandler = () => {
    if (wrapperRef.current) {
      const { scrollWidth, clientWidth } = wrapperRef.current;
      setScrollButtonVisible(scrollWidth - clientWidth > 1);
      checkScrollButtonStatus();
    }
  };
  // 容器尺寸变化（往往是受到挤压或者缩放）
  const wrapperRef = useResizeObserver<HTMLDivElement>(resizeHandler);
  // 列表尺寸变化（往往是数据变化）
  const innerListRef = useResizeObserver<HTMLDivElement>(resizeHandler);
  const composedRef = composeRefs<HTMLDivElement>(wrapperRef, ref);

  /**
   * 当 tabBar 滚动时，调整 ScrollButton 的禁用态
   */
  const checkScrollButtonStatus = React.useCallback(() => {
    if (!wrapperRef.current) {
      return;
    }
    const leftButtonDisabled = wrapperRef.current.scrollLeft === 0;
    const rightButtonDisabled =
      Math.round(wrapperRef.current.scrollLeft + wrapperRef.current.clientWidth) >=
      Math.round(wrapperRef.current.scrollWidth);
    setScrollButtonDisabled([leftButtonDisabled, rightButtonDisabled]);
  }, [wrapperRef]);

  /**
   * 从左向右滚动
   */
  const gotoNextPage = React.useCallback(() => {
    const containerRight = wrapperRef.current?.getBoundingClientRect().right || 0;
    /**
     * 找到第一个未显示完全的 tab
     */
    for (const pane of panes) {
      const tab = tabElementMapRef.current?.get(pane.key);
      if (!tab) {
        continue;
      }
      const { right: tabRight } = tab.getBoundingClientRect();
      if (tabRight > containerRight) {
        tab.scrollIntoView({
          ...defaultScrollOption,
          // 让选项滚动到最左侧，而不是仅仅滚动到可见区域
          inline: 'start',
        });
        return;
      }
    }
  }, [panes, tabElementMapRef, wrapperRef]);

  /**
   * 从右向左滚动
   */
  const gotoPrePage = React.useCallback(() => {
    const containerLeft = wrapperRef.current?.getBoundingClientRect().left || 0;
    /**
     * 找到第一个未显示完全的 tab
     */
    for (const pane of [...panes].reverse()) {
      const tab = tabElementMapRef.current?.get(pane.key);
      if (!tab) {
        continue;
      }
      const { left: tabLeft } = tab.getBoundingClientRect();
      if (tabLeft < containerLeft) {
        tab.scrollIntoView({
          ...defaultScrollOption,
          // 让选项滚动到最右侧，而不是仅仅滚动到可见区域
          inline: 'end',
        });
        return;
      }
    }
  }, [panes, tabElementMapRef, wrapperRef]);

  const getScrollButtonSx = (position: 'left' | 'right') => [
    type !== 'capsule' &&
      stylex.create(
        {
          display: 'flex',
          alignSelf: 'stretch',
          alignItems: 'center',
          width: 20,
          fontSize: 12,
          cursor: 'pointer',
          color: $token['icon-n1'],
          position: 'relative',
          background: $token['udtoken-component-bg-right-mask'],
        },
        rootCssScope('__scroll-button'),
      ),
    (position === 'left' ? scrollButtonDisabled[0] : scrollButtonDisabled[1]) &&
      stylex.create({
        cursor: 'not-allowed',
        color: $token['icon-disabled'],
      }),
  ];

  /**
   * 选中 tab,将 tabBar 滚动到可视区域内
   */
  const scrollToSelectedTab = React.useCallback(
    (key: string) => {
      const tab = tabElementMapRef.current?.get(key);
      const wrapperEle = wrapperRef.current;
      if (!tab || !wrapperEle) {
        return;
      }
      const { right: containerRight, left: containerLeft } = wrapperEle.getBoundingClientRect();
      const { right: tabRight, left: tabLeft } = tab.getBoundingClientRect();

      // 在左侧
      if (tabLeft < containerLeft + scrollMaskWidth) {
        wrapperEle.scroll({
          // 此时可能出现负数，但浏览器能兼容
          left: tab.offsetLeft - scrollMaskWidth,
          ...defaultScrollOption,
        });
        return;
      }

      // 在右侧
      if (tabRight > containerRight - scrollMaskWidth) {
        wrapperEle.scroll({
          left: wrapperEle.scrollLeft + (tabRight - containerRight) + scrollMaskWidth,
          ...defaultScrollOption,
        });
        return;
      }
    },
    [scrollMaskWidth, tabElementMapRef, wrapperRef],
  );
  return (
    <div
      style={{ columnGap: scrollButtonGutter }}
      sx={[
        stylex.create({
          display: 'flex',
          alignItems: 'center',
          width: '100%',
        }),
      ]}
    >
      {Boolean(scrollButtonVisible) && (
        <div sx={getScrollButtonSx('left')} onClick={scrollButtonDisabled[0] ? undefined : gotoPrePage}>
          <IconButton size="extra-small" disabled={scrollButtonDisabled[0]} icon={<LeftOutlined />} />
        </div>
      )}
      <div
        style={{ '--scroll-mask-width': scrollMaskWidth + 'px' } as React.CSSProperties}
        sx={[
          stylex.create({
            position: 'relative',
            minWidth: 0,
            flexGrow: 1,
            flexShrink: 1,
          }),
          !scrollButtonDisabled[0] &&
            stylex.create({
              ':before': {
                content: '""',
                position: 'absolute',
                pointerEvents: 'none',
                zIndex: 2,
                left: 0,
                top: 0,
                bottom: 6,
                width: 'var(--scroll-mask-width)',
                display: 'block',
                background: $token['udtoken-component-bg-left-mask'],
              },
            }),
          !scrollButtonDisabled[1] &&
            stylex.create({
              ':after': {
                content: '""',
                position: 'absolute',
                pointerEvents: 'none',
                zIndex: 2,
                right: 0,
                top: 0,
                bottom: 6,
                width: 'var(--scroll-mask-width)',
                display: 'block',
                background: $token['udtoken-component-bg-right-mask'],
              },
            }),
        ]}
      >
        <div
          ref={composedRef}
          className={rootCssScope('__tab-bar-wrap')}
          onScroll={checkScrollButtonStatus}
          sx={[
            stylex.create({
              overflowX: 'auto',
              position: 'relative',
              scrollbarWidth: 'none',
              '& ::-webkit-scrollbar': {
                height: 0,
              },
            }),
          ]}
        >
          <TabsInkBar offsetY={inkOffsetY} />
          <div
            ref={innerListRef}
            sx={[
              stylex.create({
                display: 'flex',
                flexWrap: 'nowrap',
                position: 'relative',
                flexShrink: 1,
                flexGrow: 0,
              }),
            ]}
          >
            {panes.map(pane => (
              <div key={pane.key} style={resolveSizingBehavior(rootTabsSizingBehavior)}>
                <TabProvider tabKey={pane.key}>
                  <Tab onClick={() => scrollToSelectedTab(pane.key)} />
                </TabProvider>
              </div>
            ))}
          </div>
        </div>
      </div>
      {Boolean(scrollButtonVisible) && (
        <div sx={getScrollButtonSx('right')} onClick={scrollButtonDisabled[1] ? undefined : gotoNextPage}>
          <IconButton size="extra-small" disabled={scrollButtonDisabled[1]} icon={<RightOutlined />} />
        </div>
      )}
    </div>
  );
});
