// @universe-design/react/src/tabs/sizing_behavior.ts
import { isNonNullable } from '../util';

/**
 * 共享的尺寸约束行为接口
 * 定义了所有尺寸模式都支持的基础约束属性
 */
export interface SharedSizingBehavior {
  /** 最小宽度约束 */
  minWidth?: number | string;
  /** 最大宽度约束 */
  maxWidth?: number | string;
}

/**
 * 弹性尺寸行为接口
 * 基于 flex 布局的比例分配模式
 *
 *  @docgen
 */
export interface FlexSizingBehavior extends SharedSizingBehavior {
  /** 尺寸模式：弹性布局 */
  mode: 'flex';
  /** flex 比例值，决定元素在可用空间中的分配比例 */
  ratio?: number;
}

/**
 * 固定尺寸行为接口
 * 设置元素为固定宽度，不参与弹性布局
 *
 *  @docgen
 */
export interface FixedSizingBehavior extends SharedSizingBehavior {
  /** 尺寸模式：固定宽度 */
  mode: 'fixed';
  /** 固定的宽度值 */
  width: number | string;
}

/**
 * 内容包裹尺寸行为接口
 * 元素宽度根据内容自适应（类似 fit-content）
 *
 *  @docgen
 */
export interface HugSizingBehavior extends SharedSizingBehavior {
  /** 尺寸模式：内容包裹 */
  mode: 'hug';
}

/**
 * 尺寸行为联合类型
 * 支持四种不同的尺寸控制模式
 */
export type SizingBehavior = FlexSizingBehavior | FixedSizingBehavior | HugSizingBehavior;

/**
 * 将尺寸行为配置转换为 CSS 样式属性
 * @param sizing - 尺寸行为配置对象
 * @returns 对应的 CSS 样式属性
 */
export function resolveSizingBehavior(sizingBehavior: SizingBehavior): React.CSSProperties {
  // 基础样式：包含盒模型和尺寸约束
  const base: React.CSSProperties = {
    boxSizing: 'border-box', // 确保 padding 和 border 包含在宽度内
    ...(isNonNullable(sizingBehavior.minWidth)
      ? {
          minWidth: sizingBehavior.minWidth, // 应用最小宽度约束
        }
      : {}),
    maxWidth: sizingBehavior.maxWidth, // 应用最大宽度约束
  };

  switch (sizingBehavior.mode) {
    case 'fixed':
      // 固定宽度模式：设置具体宽度值，不参与 flex 布局
      return { ...base, width: sizingBehavior.width, flexGrow: 0, flexShrink: 0, flexBasis: 'auto' };
    case 'hug':
      // 内容包裹模式：宽度适应内容，不参与 flex 布局
      return { ...base, width: 'fit-content', flexGrow: 0, flexShrink: 0, flexBasis: 'auto' };
    case 'flex': {
      // 弹性布局模式：根据比例分配可用空间
      const ratio = sizingBehavior.ratio ?? 1;
      // flexBasis: 0 是为了能正常缩小
      return { minWidth: 0, ...base, flexGrow: ratio, flexShrink: 1, flexBasis: 0 };
    }
    default:
      // 类型安全检查：确保所有模式都被处理
      throw new Error(`Unknown sizing mode: ${(sizingBehavior as any).mode}`);
  }
}
