// @universe-design/react/src/tabs/sx.ts
/**
 * Stylex CSS
 */
import type { StylexInstance } from '@universe-design/stylex';
import stylex from '@universe-design/stylex';
import type { TabsContextType, TabsType } from './TabsContext';
import {
  cssTypographyBody0,
  cssTypographyBody2,
  cssTypographyTitle4,
  cssTypographyTitle5,
  cssTypographyHeadline,
  cssTypographyCaption0,
} from '../stylex';
import { makeStylexStyles } from '@universe-design/stylex/react';

// Tab 的版面样式
export const sxTabTypography = (size: TabsContextType['size'], isSelected: boolean) => {
  if (size === 'large') {
    if (isSelected) {
      return cssTypographyTitle4;
    }
    return cssTypographyTitle5;
  }
  if (size === 'medium') {
    if (isSelected) {
      return cssTypographyHeadline;
    }
    return cssTypographyBody0;
  }
  if (isSelected) {
    return cssTypographyCaption0;
  }
  return cssTypographyBody2;
};

export const sxTabSize = (type: TabsType, size: TabsContextType['size']) => {
  const tabSizes: {
    [type in TabsType]?: {
      [size in TabsContextType['size'] | 'base']?: StylexInstance<any>;
    };
  } = {
    tag: {
      large: stylex.create({
        padding: '4px 12px',
        marginRight: 12,
      }),
      medium: stylex.create({
        padding: '3px 8px',
        marginRight: 12,
      }),
      small: stylex.create({
        padding: '2px 8px',
        marginRight: 8,
      }),
    },
    line: {
      small: stylex.create({
        paddingTop: 8,
        paddingBottom: 8,
      }),
    },
    card: {
      base: stylex.create({
        margin: 0,
        position: 'relative',
        zIndex: 1,
      }),

      large: stylex.create({
        height: 40,
        padding: '0 20px',
      }),
      medium: stylex.create({
        height: 36,
        padding: '0 20px',
      }),
      small: stylex.create({
        height: 32,
        padding: '0 20px',
      }),
    },
    capsule: {
      large: stylex.create({
        height: 36,
        padding: '6px 12px',
      }),
      medium: stylex.create({
        height: 32,
        padding: '5px 12px',
      }),
      small: stylex.create({
        height: 28,
        padding: '4px 12px',
      }),
    },
  };

  const sxTabBasic = stylex.create({
    paddingTop: 12,
    paddingBottom: 12,
    justifyContent: 'center',
  });
  return stylex.compose(sxTabBasic, tabSizes[type]?.base, tabSizes[type]?.[size]);
};

export const useTabStyles = makeStylexStyles(({ prefix }: { prefix: string }) => ({
  tabHover: stylex.create({
    ':hover:not(&--disabled)': {
      color: $token['primary-content-default'],
    },
  }),
  cardBase: stylex.create({
    borderWidth: 1,
    borderStyle: 'solid',
  }),
  cardDefault: stylex.create(
    {
      borderColor: 'transparent',
    },
    `${prefix}__tab--card-with-border`,
  ),
  cardActive: stylex.create(
    {
      borderColor: $token['line-border-card'],
      borderBottomColor: 'transparent',
      borderTopLeftRadius: $token['border-radius-sm'],
      borderTopRightRadius: $token['border-radius-sm'],
      backgroundColor: $token['bg-base'],
      zIndex: 1,
    },
    `${prefix}__tab--card-with-border--active`,
  ),
  capsuleBase: stylex.create({
    color: $token['text-caption'],
    borderWidth: 1,
    borderStyle: 'solid',
    borderRadius: $token['border-radius-sm'],
  }),
  capsuleDefault: stylex.create(
    {
      borderColor: 'transparent',

      ':hover': {
        backgroundColor: 'var(--udtoken-tabs-view-bg-hover)',
      },
    },
    `${prefix}__tab--capsule-with-border`,
  ),
  capsuleActive: stylex.create(
    {
      backgroundColor: $token['udtoken-component-outlined-bg'],
      borderColor: $token['line-border-component'],
      color: $token['primary-content-default'],
    },
    `${prefix}__tab--capsule-with-border--active`,
  ),
}));

export const useTabsMoreStyles = makeStylexStyles(({ prefix }: { prefix: string }) => ({
  capsuleBase: stylex.create(
    {
      color: $token['text-caption'],
    },
    `${prefix}__tab-bar-more__capsule`,
  ),
  capsuleHovered: stylex.create({
    ':hover': {
      backgroundColor: 'var(--udtoken-tabs-view-bg-hover)',
      borderRadius: $token['border-radius-md'],
    },
  }),
  capsulePressed: stylex.create(
    {
      backgroundColor: $token['udtoken-component-outlined-bg'],
      borderWidth: 1,
      borderStyle: 'solid',
      borderColor: $token['line-border-component'],
      borderRadius: $token['border-radius-md'],
      color: $token['primary-content-default'],
    },
    `${prefix}__tab-bar-more-capsule--active`,
  ),
}));
