// @universe-design/react/src/tabs/TabInk.tsx
import clsx from 'clsx';
import type { CSSProperties } from 'react';
import React, { useRef, useCallback, useState } from 'react';
import { useDebouncedCallback } from '../hooks';
import { useOverflowContext } from '../overflow/hooks';
import { useTabsContext } from './TabsContext';

function getOffsetRect(el: HTMLElement | undefined | null) {
  const { offsetLeft = 0, offsetTop = 0, offsetHeight = 0, offsetWidth = 0 } = el ?? {};
  return {
    left: offsetLeft,
    top: offsetTop,
    height: offsetHeight,
    width: offsetWidth,
    bottom: offsetTop + offsetHeight,
    right: offsetLeft + offsetWidth,
  };
}

/**
 * Tab 标签下划线
 */
export const TabsInkBar = ({
  offsetY,
}: {
  /**
   * 标签下划线的偏移量，是 divider 的高度，因为我们的 divider 是 div 元素实现的而非 border 等
   */
  offsetY: number;
}) => {
  const { type, rootCssScope, activeKey, panes, tabElementMapRef, tabBarWrapperRef, moreTabRef } = useTabsContext();
  const { omittedKeys, itemWidths, restWidth, containerWidth } = useOverflowContext();
  const inkRef = useRef<HTMLDivElement>(null);
  const getInkStyle = useCallback((): CSSProperties => {
    // overflow item 每个元素的变化都会触发这个元素 rerender
    const tabBarWrapper = tabBarWrapperRef.current;
    if (!tabBarWrapper) {
      return {};
    }

    const inkOffsetRect = getOffsetRect(inkRef.current);

    for (const pane of panes) {
      if (omittedKeys.includes(pane.key)) {
        const moreTab = moreTabRef.current;
        const moreOffsetRect = getOffsetRect(moreTab);
        return {
          width: moreOffsetRect.width,
          left: moreOffsetRect.left,
          // 多减去 1px 是为了（蓝色）选中线在（灰色）分割线上面紧贴合
          top: moreOffsetRect.bottom - inkOffsetRect.height + offsetY,
        };
      }

      if (pane.key === activeKey) {
        const tab = tabElementMapRef.current?.get(activeKey);
        const tabOffsetRect = getOffsetRect(tab);
        return {
          width: tabOffsetRect.width,
          left: tabOffsetRect.left,
          // 多减去 1px 是为了（蓝色）选中线在（灰色）分割线上面紧贴合
          top: tabOffsetRect.bottom - inkOffsetRect.height + offsetY,
        };
      }
    }
    return {};
  }, [activeKey, moreTabRef, offsetY, omittedKeys, panes, tabBarWrapperRef, tabElementMapRef]);
  const [inkStyle, setInkStyle] = useState<CSSProperties>(getInkStyle);

  /**
   * 保证在处于稳定态的时候才加渐变动画，这样默认的 activeKey 不再最左侧时，首次加载不会有一个飞过去的效果
   */
  const [animated, setAnimated] = useState<boolean>(false);
  const debouncedSetAnimated = useDebouncedCallback(
    useCallback(() => {
      setAnimated(true);
    }, []),
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    800,
  );

  React.useEffect(() => {
    // 为了减少一些不必要的 resize observer 所以这里直接通过 overflow item 宽度变换来代替真实元素的尺寸变化
    itemWidths;
    restWidth;
    containerWidth;
    const newStyle = getInkStyle();
    /**
     * 如果使用 display: none 控制元素显影，不重置 style
     * 防止其位置抖动
     **/
    if (newStyle.width === 0) {
      return;
    }

    setInkStyle(newStyle);
    debouncedSetAnimated();
  }, [containerWidth, debouncedSetAnimated, getInkStyle, itemWidths, restWidth, tabBarWrapperRef]);

  if (type === 'line') {
    const inkCls = clsx(rootCssScope('__tab-bar-ink'), animated && rootCssScope('__tab-bar-ink-animated'));
    return <div style={inkStyle} className={inkCls} ref={inkRef} />;
  }
  return null;
};
