// @universe-design/react/src/tabs/Tabs.tsx
/* eslint-disable react/no-unknown-property */
/** @jsxRuntime classic */
/** @jsx jsx */
/** @jsxFrag React.Fragment */
import stylex from '@universe-design/stylex';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx } from '@universe-design/stylex/react';
import DownBoldOutlined from '@universe-design/icons-react/DownBoldOutlined';
import { cssTypographyTitle4, cssTypographyTitle5 } from '../stylex';
import clsx from 'clsx';
import type { ReactNode } from 'react';
import React, { forwardRef, useImperativeHandle, useMemo, useRef, useState } from 'react';
import { useLocaleReceiver } from '../config-provider/hooks';
import type { DropdownProps } from '../dropdown';
import { Dropdown } from '../dropdown';
import { Menu } from '../menu';
import { Overflow } from '../overflow';
import { useOverflowContext } from '../overflow/hooks';
import { getDefaultSizeAbbr, isFunction, __SCREENSHOT__ } from '../util';
import { TabPane } from './TabPane';
import type { TabsProps } from './TabsContext';
import { TabsContext, useTabs, useTabsContext } from './TabsContext';
import { sxTabSize, sxTabTypography, useTabsMoreStyles } from './sx';
import { Tab } from './Tab';
import { TabProvider } from './TabContext';
import { ScrollableTabsHeader } from './ScrollableTabsHeader';
import { TabsInkBar } from './TabInk';
import './style';
import { resolveSizingBehavior } from './sizing_behavior';

export const TabMoreArrow = () => <DownBoldOutlined />;

export const TabMoreContent = () => {
  const locale = useLocaleReceiver('tabs');
  return <span>{locale.more}</span>;
};

/**
 * `<TabMore />` 组件的属性
 * @public
 */
export interface TabMoreProps extends Omit<DropdownProps, 'overlay' | 'children' | 'visible' | 'onVisibleChange'> {
  /**
   * 子节点
   */
  children?: React.ReactNode | ((ctx: { omittedKeys: string[] }) => React.ReactNode);
}

export const TabMore = ({ children, ...dropDownProps }: TabMoreProps) => {
  const { panes, rootCssScope, handleTabClick, activeKey, moreTabRef, size, type } = useTabsContext();
  const { omittedKeys } = useOverflowContext();
  const [showMoreTabs, setShowMoreTabs] = useState<boolean>(false);
  const isSelected = omittedKeys.includes(activeKey);
  const mergedVisible = __SCREENSHOT__ ? isSelected : showMoreTabs && Boolean(omittedKeys.length);
  const tabsMoreStyles = useTabsMoreStyles({ prefix: rootCssScope() });

  const morePanes = panes.filter(p => omittedKeys.includes(p.key));
  const cls = clsx(rootCssScope('__tab-bar-more'), rootCssScope(`__tab-bar-more--${type}`), {
    [rootCssScope('__tab-bar-more--selected')]: isSelected,
    [rootCssScope('__tab-bar-more--show')]: mergedVisible,
  });

  const menu = (
    <Menu selectedKeys={[activeKey]} textOverflow={false} selectable>
      {morePanes.map(pane => {
        const { disabled, key, tab } = pane;
        return (
          <Menu.Item key={key} disabled={disabled} onClick={({ domEvent }) => handleTabClick(key, domEvent)}>
            <TabProvider tabKey={key}>{tab}</TabProvider>
          </Menu.Item>
        );
      })}
    </Menu>
  );

  const moreTabContentNode = (
    <>
      <TabMoreContent />
      <TabMoreArrow />
    </>
  );

  return (
    <div className={clsx(rootCssScope('__tab-bar__operation'))}>
      <div className={cls} sx={[!isSelected ? cssTypographyTitle5 : cssTypographyTitle4]}>
        <Dropdown
          trigger={['hover', 'click']}
          {...dropDownProps}
          overlay={menu}
          visible={mergedVisible}
          onVisibleChange={setShowMoreTabs}
        >
          <div
            ref={moreTabRef}
            className={clsx(
              rootCssScope('__tab-bar-more-body'),
              type === 'tag' && [
                rootCssScope('-sm-border-radius'),
                isSelected
                  ? [rootCssScope('-primary-background'), rootCssScope('-color-pri-text')]
                  : rootCssScope('-neutral-background'),
              ],
            )}
            sx={[
              sxTabTypography(size, isSelected),
              sxTabSize(type, size),
              type === 'capsule' && [
                tabsMoreStyles.capsuleBase,
                !isSelected && tabsMoreStyles.capsuleHovered,
                isSelected && tabsMoreStyles.capsulePressed,
              ],
            ]}
          >
            {(isFunction(children)
              ? children({
                  omittedKeys: omittedKeys as string[],
                })
              : children) ?? moreTabContentNode}
          </div>
        </Dropdown>
      </div>
    </div>
  );
};

export const Tabs = forwardRef<HTMLDivElement, TabsProps>((props, ref) => {
  const { tabMore, sticky } = props;
  const tabCtx = useTabs(props);
  const {
    style,
    className,
    tabBarStyle,
    type = 'line',
    size = 'medium',
    maxCount,
    tabBarExtraContent,
    rootCssScope,
    tabBarWrapperRef,
    panes,
    rootTabsSizingBehavior,
    scrollable,
  } = tabCtx;

  const usedSize = getDefaultSizeAbbr(size, 'medium');
  const tabCls = clsx(
    rootCssScope(),
    {
      [rootCssScope(`--type-${type}`)]: type,
      [rootCssScope(`--size-${usedSize}`)]: usedSize,
    },
    className,
  );

  const tabsRef = useRef<HTMLDivElement>(null);
  // 蓝色选中线和 divider 重合
  const inkOffsetY = 0;

  useImperativeHandle<HTMLDivElement | null, HTMLDivElement | null>(ref, () => tabsRef.current);

  const renderPanes = () => (
    <div className={rootCssScope('__content')}>
      {panes.map(paneProps => (
        <TabPane {...paneProps} internalKey={paneProps.key} key={paneProps.key} />
      ))}
    </div>
  );

  const tabsInkBarNode = useMemo(() => <TabsInkBar offsetY={inkOffsetY} />, [inkOffsetY]);
  const moreTabNode = useMemo(() => tabMore ?? <TabMore />, [tabMore]);

  const renderLeftExtraContentNode = useMemo((): ReactNode => {
    const assertExtra = tabBarExtraContent as any;
    if (!assertExtra) {
      return null;
    }
    return <div className={rootCssScope('__tab-extra-content')}>{assertExtra.left}</div>;
  }, [rootCssScope, tabBarExtraContent]);

  const renderRightExtraContentNode = useMemo((): ReactNode => {
    const assertExtra = tabBarExtraContent as any;
    if (!assertExtra) {
      return null;
    }
    return (
      <div className={rootCssScope('__tab-extra-content')}>
        {assertExtra.right || (!assertExtra.left && assertExtra) || null}
      </div>
    );
  }, [rootCssScope, tabBarExtraContent]);

  const getItemStyle = React.useCallback(() => resolveSizingBehavior(rootTabsSizingBehavior), [rootTabsSizingBehavior]);

  const renderTabBar = () => (
    <div
      className={clsx(
        rootCssScope('__tab-bar-holder'),
        rootCssScope('__tab-bar-nav'),
        sticky && rootCssScope('__tab-bar-nav--sticky'),
      )}
      sx={[
        stylex.create({
          position: 'relative',
        }),
        sticky &&
          stylex.create({
            position: 'sticky',
            top: 0,
            backgroundColor: $token['bg-body'],
          }),
        type === 'capsule' &&
          stylex.create({
            background: $token['N900-5'],
            borderRadius: $token['border-radius-sm'],
          }),
        // scrollable 时会撑开父级，无法滚动 && responsive 时无法正常展开，只有一个 more
        type === 'capsule' && !scrollable && maxCount !== 'responsive' && stylex.create({ width: 'fit-content' }),
      ]}
    >
      {renderLeftExtraContentNode}
      {scrollable ? (
        <ScrollableTabsHeader ref={tabBarWrapperRef} inkOffsetY={inkOffsetY} />
      ) : (
        <div ref={tabBarWrapperRef} className={rootCssScope('__tab-bar-wrap')}>
          <Overflow<string>
            style={tabBarStyle}
            getItemStyle={getItemStyle}
            restStyle={resolveSizingBehavior(rootTabsSizingBehavior)}
            keys={panes.map(p => p.key)}
            maxCount={maxCount}
            itemRenderer={key => (
              <TabProvider tabKey={key}>
                <Tab />
              </TabProvider>
            )}
            restRenderer={() => moreTabNode}
            addonRenderer={() => tabsInkBarNode}
            alwaysObserveResize
          />
        </div>
      )}

      {renderRightExtraContentNode}
      {type === 'line' && <div className={rootCssScope('__tab-bar-holder-divider')} />}
    </div>
  );
  return (
    <TabsContext.Provider value={tabCtx}>
      <div ref={tabsRef} style={style} className={tabCls}>
        {renderTabBar()}
        {renderPanes()}
      </div>
    </TabsContext.Provider>
  );
});

if (__DEV__) {
  Tabs.displayName = 'Tabs';
}
