// @universe-design/react/src/tabs/TabsContext.ts
import type { CSSProperties, KeyboardEvent, ReactNode, MouseEvent, RefObject, ReactElement } from 'react';
import * as React from 'react';
import { createContext, useContext } from 'use-context-selector';
import type { Required } from 'utility-types';
import { useControlledState } from '@universe-design/react-utilities';
import { useCssScope } from '../config-provider/hooks';
import type { DefaultSizeType } from '../util';
import { isFunction, toFlatArray } from '../util';
import type { SizingBehavior } from './sizing_behavior';

/**
 * Tab 尺寸行为类型
 * 定义单个 Tab 的尺寸控制方式
 * 支持简单字符串类型和详细的 SizingBehavior 对象类型
 */
export type TabSizingBehavior = 'flex' | 'hug' | SizingBehavior;

export const UD_TABPANE_ID = 'UD_TABPANE_ID';
export function isTabPane(node: any): node is React.ReactElement<TabPaneProps> {
  return node?.type?.[UD_TABPANE_ID];
}

export type TabPosition = 'left' | 'right' | 'top' | 'bottom';

/**
 * @public
 */
export type TabBarExtraContent = ReactNode | { left: ReactNode; right: ReactNode };

export type TabsType = 'line' | 'card' | 'tag' | 'capsule'; // 'editable-card' 类型暂缓提供

export type onEditFunction = (action: 'add' | 'delete', targetKey?: string) => void;

/**
 * `<Tabs />` 组件的属性
 * @public
 */
export interface TabsProps {
  /**
   * 节点样式
   */
  style?: CSSProperties;
  /**
   * 节点类名
   */
  className?: string;
  /**
   * 页签激活 `key` 的受控值
   */
  activeKey?: string;
  /**
   * 页签激活 `key` 的默认值
   */
  defaultActiveKey?: string;
  /**
   * 页签条左右的额外内容元素
   */
  tabBarExtraContent?: TabBarExtraContent;
  /**
   * 页签之间的间隙
   * @defaultValue `tag`: 16, `others`: 28
   */
  tabBarGutter?: number;
  /**
   * ScrollButton 和 TabBar 的间距，默认 line 类型的为 28，其他为 0
   */
  scrollButtonGutter?: number;
  /**
   * 页签条的 `style` 属性透传
   */
  tabBarStyle?: CSSProperties;
  // tabPosition?: TabPosition;
  children: ReactNode;
  /**
   * 页签类型
   * @defaultValue `line`
   */
  type?: TabsType; // 'editable-card' 类型暂缓提供
  /**
   * 页签大小
   * @defaultValue `medium`
   */
  size?: DefaultSizeType;
  /**
   * 最大页签数量
   * @defaultValue `responsive`
   */
  maxCount?: number | 'responsive';
  /**
   * 切换面板后触发的回调
   */
  onChange?: (activeKey: string) => void;
  /**
   * 页签被点击的回调
   */
  onTabClick?: (activeKey: string, e: KeyboardEvent | MouseEvent) => void;
  /**
   * 「更多」页签节点
   * @defaultValue `<TabMore />`
   */
  tabMore?: ReactElement;

  /**
   * 双击事件
   */
  onTabDoubleClick?: (activeKey: string, e: KeyboardEvent | MouseEvent) => void;
  /**
   * 当页签过多时，是否横向滚动。默认收起到【更多】
   * @defaultValue false
   */
  scrollable?: boolean;
  /**
   * 页面滚动时，TabBar 是否固定在顶部
   * @defaultValue false
   */
  sticky?: boolean;
  /**
   * 统一设置 tabs 中每个 tab 的 sizingBehavior
   * 支持简单字符串类型：
   * - `flex`: 弹性布局，平均分配空间
   * - `hug`: 内容包裹，根据内容自适应宽度
   * 也支持 SizingBehavior 对象类型，可以精确控制 minWidth、maxWidth 等属性
   * @defaultValue `hug`
   */
  tabsSizingBehavior?: TabSizingBehavior;
}

/**
 * `<Tabs.Pane>` 组件的属性
 * @public
 */
export interface TabPaneProps {
  /**
   * 窗格样式
   */
  style?: CSSProperties;
  /**
   * 窗格类名
   */
  className?: string;
  /**
   * 选项卡头显示文字
   */
  tab?: ReactNode;
  /**
   * 对应`activeKey`
   */
  key?: string;
  /**
   * 禁用窗格
   * @defaultValue false
   */
  disabled?: boolean;
  /**
   * 是否将隐藏的 Pane 也提前渲染出来
   * @defaultValue false
   */
  forceRender?: boolean;
  /**
   * 激活态的 Pane 被切换之后，是否将其销毁， 默认不销毁。
   * @defaultValue false
   */
  destroyOnHide?: boolean;
  internalKey?: string;
  children?: ReactNode;
}

export interface TabsContextType
  extends Required<
    Omit<TabsProps, 'defaultActiveKey' | 'tabsSizingBehavior'>,
    'type' | 'size' | 'maxCount' | 'activeKey' | 'tabBarGutter' | 'scrollable'
  > {
  rootCssScope: (suffix?: string) => string;
  setActiveKey: (k: string) => void;
  handleTabClick: (key: string, e: MouseEvent<HTMLElement> | KeyboardEvent<HTMLElement>) => void;
  handleTabDoubleClick: (key: string, e: MouseEvent<HTMLElement> | KeyboardEvent<HTMLElement>) => void;
  panes: Required<TabPaneProps, 'key'>[];
  activeTabIndex: number;
  tabElementMapRef: RefObject<Map<string, HTMLDivElement | null>>;
  tabBarWrapperRef: RefObject<HTMLDivElement>;
  moreTabRef: RefObject<HTMLDivElement>;
  rootTabsSizingBehavior: SizingBehavior;
}

export const TabsContext = createContext<TabsContextType>({} as any);

if (__DEV__) {
  TabsContext.displayName = 'TabsContext';
}

export const useTabs = (props: TabsProps): TabsContextType => {
  const {
    style,
    className,
    children,
    type = 'line',
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    tabBarGutter = type === 'tag' ? 16 : 28,
    scrollButtonGutter = type === 'line' ? 28 : 0,
    tabBarStyle,
    activeKey: actKeyInProps,
    defaultActiveKey,
    size = 'medium',
    maxCount: maxCountProp,
    tabBarExtraContent,
    onChange,
    onTabClick,
    onTabDoubleClick,
    scrollable: scrollableProp,
    tabsSizingBehavior,
  } = props;

  const maxCount = (() => {
    if (!maxCountProp && type === 'capsule') {
      return Infinity;
    }
    if (!maxCountProp) {
      return 'responsive';
    }
    return maxCountProp;
  })();

  const scrollable = (() =>
    // if (type === 'capsule') {
    //   warning(!scrollableProp, devWarning('Tabs', 'scrollable is not supported in capsule type'));
    //   return false;
    // }
    scrollableProp ?? false)();

  const rootTabsSizingBehavior = React.useMemo((): SizingBehavior => {
    if (!tabsSizingBehavior || tabsSizingBehavior === 'hug') {
      return {
        mode: 'hug',
        maxWidth: 240,
      };
    }
    if (tabsSizingBehavior === 'flex') {
      return {
        mode: 'flex',
        maxWidth: 240,
      };
    }
    return tabsSizingBehavior;
  }, [tabsSizingBehavior]);

  const panes = React.useMemo(
    () =>
      toFlatArray(children)
        .filter(isTabPane)
        .map((child, index) => ({
          ...child.props,
          key: (child.key && String(child.key)) || `${index}`,
        })),
    [children],
  );

  const [activeKey, setActiveKey] = useControlledState<string>(defaultActiveKey || panes[0].key, actKeyInProps);

  const handleTabClick = React.useCallback<TabsContextType['handleTabClick']>(
    (key, e) => {
      const pane = panes.find(p => p.key === key);
      if (!pane || pane.disabled) {
        return;
      }
      setActiveKey(key);
      isFunction(onChange) && onChange(key);
      isFunction(onTabClick) && onTabClick(key, e);
    },
    [onChange, onTabClick, panes, setActiveKey],
  );

  const handleTabDoubleClick = React.useCallback<TabsContextType['handleTabDoubleClick']>(
    (key, e) => {
      const pane = panes.find(p => p.key === key);
      if (!pane || pane.disabled) {
        return;
      }
      if (isFunction(onTabDoubleClick)) {
        onTabDoubleClick(key, e);
      }
    },
    [onTabDoubleClick, panes],
  );

  const rootCssScope = useCssScope('tabs');

  const activeTabIndex = panes.findIndex(p => p.key === activeKey);

  const tabElementMapRef = React.useRef(new Map());
  // position relative 的元素，用于定位 ink
  const tabBarWrapperRef = React.useRef<HTMLDivElement>(null);
  // 更多 tab
  const moreTabRef = React.useRef<HTMLDivElement>(null);

  return React.useMemo(
    () => ({
      style,
      className,
      children,
      tabBarGutter,
      tabBarStyle: {
        // avoid browser's text selection behavior, when double click
        userSelect: isFunction(onTabDoubleClick) ? 'none' : 'auto',
        ...tabBarStyle,
        ...(scrollable
          ? {
              flexWrap: 'nowrap',
            }
          : {}),
      },
      activeKey,
      type,
      size,
      maxCount,
      tabBarExtraContent,
      onChange,
      onTabClick,
      onTabDoubleClick,
      rootCssScope,
      setActiveKey,
      handleTabClick,
      handleTabDoubleClick,
      panes,
      activeTabIndex,
      tabElementMapRef,
      tabBarWrapperRef,
      moreTabRef,
      scrollButtonGutter,
      rootTabsSizingBehavior,
      scrollable,
    }),
    [
      activeKey,
      activeTabIndex,
      children,
      className,
      handleTabClick,
      handleTabDoubleClick,
      maxCount,
      onChange,
      onTabClick,
      onTabDoubleClick,
      panes,
      rootCssScope,
      scrollButtonGutter,
      scrollable,
      setActiveKey,
      size,
      style,
      tabBarExtraContent,
      tabBarGutter,
      tabBarStyle,
      type,
      rootTabsSizingBehavior,
    ],
  );
};

export const useTabsContext = () => useContext(TabsContext);
