// @universe-design/react/es/stylex/border-radius.sx.js
//.u55qycf{border-top-left-radius:999999px;}
import u1m59yg4 from "./../.stylex/u1m59yg4";
//.u2ie9lg{border-top-right-radius:999999px;}
import ubm82nm from "./../.stylex/ubm82nm";
//.u1ewtp5v{border-bottom-right-radius:999999px;}
import u1qcfi6m from "./../.stylex/u1qcfi6m";
//.u1j6dxmp{border-bottom-left-radius:999999px;}
import u1pc15gj from "./../.stylex/u1pc15gj";
//.ugv9nof{border-top-left-radius:12px;}
import uacbfjb from "./../.stylex/uacbfjb";
//.u24j1xx{border-top-right-radius:12px;}
import u1l86f2g from "./../.stylex/u1l86f2g";
//.u1ig3etr{border-bottom-right-radius:12px;}
import u1r1mxth from "./../.stylex/u1r1mxth";
//.uy0y5fd{border-bottom-left-radius:12px;}
import ul6wel9 from "./../.stylex/ul6wel9";
//.umb70yw{border-top-left-radius:8px;}
import u1j8h7k8 from "./../.stylex/u1j8h7k8";
//.u1spoy8{border-top-right-radius:8px;}
import u10e4wt4 from "./../.stylex/u10e4wt4";
//.u1erghxr{border-bottom-right-radius:8px;}
import u5rwohh from "./../.stylex/u5rwohh";
//.u1ehz9de{border-bottom-left-radius:8px;}
import u14kz1ee from "./../.stylex/u14kz1ee";
//.uifjimi{border-top-left-radius:6px;}
import u1n69reo from "./../.stylex/u1n69reo";
//.u8rsvsd{border-top-right-radius:6px;}
import u1sx1i35 from "./../.stylex/u1sx1i35";
//.u18fiblo{border-bottom-right-radius:6px;}
import ui28ht5 from "./../.stylex/ui28ht5";
//.u1asnnh4{border-bottom-left-radius:6px;}
import uyycwk0 from "./../.stylex/uyycwk0";
//.u8wkphi{border-top-left-radius:4px;}
import uzvwtxa from "./../.stylex/uzvwtxa";
//.uwsfkhu{border-top-right-radius:4px;}
import u5dwexa from "./../.stylex/u5dwexa";
//.uff7au0{border-bottom-right-radius:4px;}
import ukikfbv from "./../.stylex/ukikfbv";
//.u1bjk9e1{border-bottom-left-radius:4px;}
import uhf7ou4 from "./../.stylex/uhf7ou4";
//.u1127022{border-top-left-radius:2px;}
import u1ubg0fh from "./../.stylex/u1ubg0fh";
//.u1stktl2{border-top-right-radius:2px;}
import u19y2ap9 from "./../.stylex/u19y2ap9";
//.u2i1086{border-bottom-right-radius:2px;}
import u16wxydm from "./../.stylex/u16wxydm";
//.u1rlsnju{border-bottom-left-radius:2px;}
import u5pet9v from "./../.stylex/u5pet9v";
import stylex from '@universe-design/stylex';
export const sxBorderRadius = {
  xxs: {
    "props": {
      "Btl43ni": u1ubg0fh,
      "B7oj6ja": u19y2ap9,
      "Bbmb7ep": u16wxydm,
      "Beyfa6y": u5pet9v
    }
  },
  xs: {
    "props": {
      "Btl43ni": uzvwtxa,
      "B7oj6ja": u5dwexa,
      "Bbmb7ep": ukikfbv,
      "Beyfa6y": uhf7ou4
    }
  },
  sm: {
    "props": {
      "Btl43ni": u1n69reo,
      "B7oj6ja": u1sx1i35,
      "Bbmb7ep": ui28ht5,
      "Beyfa6y": uyycwk0
    }
  },
  md: {
    "props": {
      "Btl43ni": u1j8h7k8,
      "B7oj6ja": u10e4wt4,
      "Bbmb7ep": u5rwohh,
      "Beyfa6y": u14kz1ee
    }
  },
  lg: {
    "props": {
      "Btl43ni": uacbfjb,
      "B7oj6ja": u1l86f2g,
      "Bbmb7ep": u1r1mxth,
      "Beyfa6y": ul6wel9
    }
  },
  xl: {
    "props": {
      "Btl43ni": u1m59yg4,
      "B7oj6ja": ubm82nm,
      "Bbmb7ep": u1qcfi6m,
      "Beyfa6y": u1pc15gj
    }
  }
};
