// @universe-design/react/src/shared/hooks/useThemeMode.ts
import * as React from 'react';
import { getDefaultMediaMatches, useIsomorphicLayoutEffect as useLayoutEffect, useMedia } from '../../hooks';
import { isClient } from '@universe-design/react-utilities';

export type ThemeMode = 'dark' | 'light';

const PrefersColorSchemeDark = '(prefers-color-scheme: dark)';

export function getSystemThemeMode(): ThemeMode {
  return getDefaultMediaMatches(PrefersColorSchemeDark) ? 'dark' : 'light';
}

const getSetThemeMode = (mode?: ThemeMode) => {
  if (!isClient()) {
    return 'light';
  }
  if (mode !== undefined) {
    document.documentElement.dataset.udThemeMode = mode;
    return mode;
  } else {
    return document.documentElement.dataset.udThemeMode as ThemeMode;
  }
};

const useSystemThemeMode = (): ThemeMode => (useMedia(PrefersColorSchemeDark) ? 'dark' : 'light');

/**
 * 用 body data-ud-theme-mode 作为 light dark 切换的依据
 */
export const useDerivedThemeModeState = () => {
  const [state, setState] = React.useState(getSetThemeMode());
  React.useEffect(() => {
    const observer = new MutationObserver(mutations => {
      mutations.forEach(() => {
        setState(getSetThemeMode());
      });
    });
    observer.observe(document.documentElement, {
      attributes: true,
      attributeFilter: ['data-ud-theme-mode'],
    });
    return () => {
      observer.disconnect();
    };
  }, []);
  return state;
};

export const useThemeMode = (
  options: {
    defaultMode?: ThemeMode;
    followSystem?: boolean;
  } = {},
) => {
  const { defaultMode = getSystemThemeMode(), followSystem = false } = options;
  const systemThemeMode = useSystemThemeMode();
  const derivedThemeMode = useDerivedThemeModeState();
  useLayoutEffect(() => {
    if (followSystem) {
      getSetThemeMode(systemThemeMode);
    }
  }, [followSystem, systemThemeMode]);
  useLayoutEffect(() => {
    if (!followSystem) {
      getSetThemeMode(defaultMode);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  const setMode = React.useCallback(
    (_mode: ThemeMode) => {
      getSetThemeMode(followSystem ? systemThemeMode : _mode);
    },
    [followSystem, systemThemeMode],
  );
  return [derivedThemeMode, setMode] as const;
};
