// @universe-design/react/lib/table/hooks/useSorter.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = useSorter;
exports.getSortData = getSortData;
var _react = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _tooltip = require("../../tooltip");
var _ExpandDownFilled = _interopRequireDefault(require("@universe-design/icons-react/ExpandDownFilled"));
var _ExpandUpFilled = _interopRequireDefault(require("@universe-design/icons-react/ExpandUpFilled"));
var _SortZToAOutlined = _interopRequireDefault(require("@universe-design/icons-react/SortZToAOutlined"));
var _SortAToZOutlined = _interopRequireDefault(require("@universe-design/icons-react/SortAToZOutlined"));
var _utils = require("../utils");
var _interface = require("../interface");
var _button = require("../../button");
function getMultiplePriority(column) {
  if (typeof column.sorter === 'object' && typeof column.sorter.multiple === 'number') {
    return column.sorter.multiple;
  }
  return false;
}
function getSortFunction(sorter) {
  if (typeof sorter === 'function') {
    return sorter;
  }
  if (sorter && typeof sorter === 'object' && sorter.compare) {
    return sorter.compare;
  }
  return false;
}
function nextSortDirection(sortDirections, current) {
  if (!current) {
    return sortDirections[0];
  }
  return sortDirections[sortDirections.indexOf(current) + 1];
}
function collectSortStates(columns, init, pos) {
  let sortStates = [];
  function pushState(column, columnPos) {
    var _column$sortOrder;
    sortStates.push({
      column,
      key: (0, _utils.getColumnKey)(column, columnPos),
      multiplePriority: getMultiplePriority(column),
      sortOrder: (_column$sortOrder = column.sortOrder) !== null && _column$sortOrder !== void 0 ? _column$sortOrder : null
    });
  }
  (columns || []).forEach((column, index) => {
    const columnPos = (0, _utils.getColumnPos)(index, pos);
    if (column.children) {
      if ('sortOrder' in column) {
        // Controlled
        pushState(column, columnPos);
      }
      sortStates = [...sortStates, ...collectSortStates(column.children, init, columnPos)];
    } else if (column.sorter) {
      if ('sortOrder' in column) {
        // Controlled
        pushState(column, columnPos);
      } else if (init && column.defaultSortOrder) {
        // Default sorter
        sortStates.push({
          column,
          key: (0, _utils.getColumnKey)(column, columnPos),
          multiplePriority: getMultiplePriority(column),
          sortOrder: column.defaultSortOrder
        });
      }
    }
  });
  return sortStates;
}
function stateToInfo(sorterStates) {
  const {
    column,
    sortOrder
  } = sorterStates;
  return {
    column,
    order: sortOrder,
    field: column.dataIndex,
    columnKey: column.key
  };
}
function generateSorterInfo(sorterStates) {
  const list = sorterStates.filter(_ref => {
    let {
      sortOrder
    } = _ref;
    return sortOrder;
  }).map(stateToInfo);
  if (list.length <= 1) {
    return list[0] || {};
  }
  return list;
}
function getSortData(data, sortStates, childrenColumnName) {
  const innerSorterStates = sortStates.slice().sort((a, b) => b.multiplePriority - a.multiplePriority);
  const cloneData = data.slice();
  const runningSorters = innerSorterStates.filter(_ref2 => {
    let {
      column: {
        sorter
      },
      sortOrder
    } = _ref2;
    return getSortFunction(sorter) && sortOrder;
  });

  // Skip if no sorter needed
  if (!runningSorters.length) {
    return cloneData;
  }
  return cloneData.sort((record1, record2) => {
    for (let i = 0; i < runningSorters.length; i += 1) {
      const sorterState = runningSorters[i];
      const {
        column: {
          sorter
        },
        sortOrder
      } = sorterState;
      const compareFn = getSortFunction(sorter);
      if (compareFn && sortOrder) {
        const compareResult = compareFn(record1, record2, sortOrder);
        if (compareResult !== 0) {
          return sortOrder === _interface.TableSort.ASCEND ? compareResult : -compareResult;
        }
      }
    }
    return 0;
  }).map(record => {
    const subRecords = record[childrenColumnName];
    if (subRecords) {
      return {
        ...record,
        [childrenColumnName]: getSortData(subRecords, sortStates, childrenColumnName)
      };
    }
    return record;
  });
}
function useSorter(_ref3) {
  let {
    prefix,
    mergedColumns,
    onSorterChange,
    sortDirections: tableSortDirections,
    tableLocale,
    showSorterTooltip: tableShowSorterTooltip
  } = _ref3;
  const [sortStates, setSortStates] = (0, _react.useState)(collectSortStates(mergedColumns, true));
  const mergedSorterStates = (0, _react.useMemo)(() => {
    let validate = true;
    const collectedStates = collectSortStates(mergedColumns, false);
    if (!collectedStates.length) {
      return sortStates;
    }
    const validateStates = [];
    function patchStates(state) {
      if (validate) {
        validateStates.push(state);
      } else {
        validateStates.push({
          ...state,
          sortOrder: null
        });
      }
    }
    let multipleMode = null;
    collectedStates.forEach(state => {
      if (multipleMode === null) {
        patchStates(state);
        if (state.sortOrder) {
          if (state.multiplePriority === false) {
            validate = false;
          } else {
            multipleMode = true;
          }
        }
      } else if (multipleMode && state.multiplePriority !== false) {
        patchStates(state);
      } else {
        validate = false;
        patchStates(state);
      }
    });
    return validateStates;
  }, [mergedColumns, sortStates]);

  // Get render columns title required props
  const columnTitleSorterProps = (0, _react.useMemo)(() => {
    var _sortColumns$, _sortColumns$2;
    const sortColumns = mergedSorterStates.map(_ref4 => {
      let {
        column,
        sortOrder
      } = _ref4;
      return {
        column,
        order: sortOrder
      };
    });
    return {
      sortColumns,
      sortColumn: (_sortColumns$ = sortColumns[0]) === null || _sortColumns$ === void 0 ? void 0 : _sortColumns$.column,
      sortOrder: (_sortColumns$2 = sortColumns[0]) === null || _sortColumns$2 === void 0 ? void 0 : _sortColumns$2.order
    };
  }, [mergedSorterStates]);
  const triggerSorter = (0, _react.useCallback)(sortState => {
    let newSorterStates;
    if (sortState.multiplePriority === false || !mergedSorterStates.length || mergedSorterStates[0].multiplePriority === false) {
      newSorterStates = [sortState];
    } else {
      newSorterStates = [...mergedSorterStates.filter(_ref5 => {
        let {
          key
        } = _ref5;
        return key !== sortState.key;
      }), sortState];
    }
    setSortStates(newSorterStates);
    onSorterChange(generateSorterInfo(newSorterStates), newSorterStates);
  }, [mergedSorterStates, onSorterChange]);
  const transformColumns = (0, _react.useCallback)(innerColumns => {
    const injectSorter = (columns, sorterSates, pos) => columns.map((column, index) => {
      const columnPos = (0, _utils.getColumnPos)(index, pos);
      let newColumn = column;
      if (newColumn.sorter) {
        var _newColumn$showSorter;
        const sortDirections = newColumn.sortDirections || tableSortDirections;
        const showSorterTooltip = (_newColumn$showSorter = newColumn.showSorterTooltip) !== null && _newColumn$showSorter !== void 0 ? _newColumn$showSorter : tableShowSorterTooltip;
        const columnKey = (0, _utils.getColumnKey)(newColumn, columnPos);
        const sorterState = sorterSates.find(_ref6 => {
          let {
            key
          } = _ref6;
          return key === columnKey;
        });
        const sorterOrder = sorterState ? sorterState.sortOrder : null;
        const nextSortOrder = nextSortDirection(sortDirections, sorterOrder);
        let upNode = null;
        let downNode = null;
        if (sortDirections.includes(_interface.TableSort.ASCEND) && sortDirections.includes(_interface.TableSort.DESCEND)) {
          upNode = /*#__PURE__*/_react.default.createElement(_ExpandUpFilled.default, {
            className: (0, _clsx.default)(`${prefix}-column-sorter-up`, {
              active: sorterOrder === _interface.TableSort.ASCEND
            })
          });
          downNode = /*#__PURE__*/_react.default.createElement(_ExpandDownFilled.default, {
            className: (0, _clsx.default)(`${prefix}-column-sorter-down`, {
              active: sorterOrder === _interface.TableSort.DESCEND
            })
          });
        } else {
          if (sortDirections.includes(_interface.TableSort.ASCEND)) {
            upNode = /*#__PURE__*/_react.default.createElement(_button.IconButton, {
              color: sorterOrder === _interface.TableSort.ASCEND ? 'primary' : undefined,
              size: "small",
              selected: sorterOrder === _interface.TableSort.ASCEND
              // info={sorterOrder === TableSort.ASCEND}
              ,
              icon: /*#__PURE__*/_react.default.createElement(_SortAToZOutlined.default, null)
            });
          }
          if (sortDirections.includes(_interface.TableSort.DESCEND)) {
            downNode = /*#__PURE__*/_react.default.createElement(_button.IconButton, {
              color: sorterOrder === _interface.TableSort.DESCEND ? 'primary' : undefined,
              size: "small",
              selected: sorterOrder === _interface.TableSort.DESCEND
              // info={sorterOrder === TableSort.DESCEND}
              ,
              icon: /*#__PURE__*/_react.default.createElement(_SortZToAOutlined.default, null)
            });
          }
        }
        const {
          cancelSort,
          triggerAsc,
          triggerDesc
        } = tableLocale || {};
        let sortTip = cancelSort;
        if (nextSortOrder === _interface.TableSort.DESCEND) {
          sortTip = triggerDesc;
        } else if (nextSortOrder === _interface.TableSort.ASCEND) {
          sortTip = triggerAsc;
        }
        newColumn = {
          ...newColumn,
          className: (0, _clsx.default)(newColumn.className, {
            [`${prefix}-column-sort`]: sorterOrder
          }),
          title: renderProps => {
            const renderSortTitle = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, (0, _utils.renderColumnTitle)(column.title, renderProps), /*#__PURE__*/_react.default.createElement(_tooltip.Tooltip, {
              title: sortTip,
              disabled: !showSorterTooltip
            }, /*#__PURE__*/_react.default.createElement("span", {
              className: (0, _clsx.default)(`${prefix}-column-sorter`, {
                [`${prefix}-column-sorter-full`]: upNode && downNode
              })
            }, /*#__PURE__*/_react.default.createElement("span", {
              className: `${prefix}-column-sorter-inner`
            }, upNode, downNode))));
            return renderSortTitle;
          },
          onHeaderCell: col => {
            const cell = column.onHeaderCell && column.onHeaderCell(col) || {};
            const originOnClick = cell.onClick;
            cell.onClick = event => {
              triggerSorter({
                column,
                key: columnKey,
                sortOrder: nextSortOrder,
                multiplePriority: getMultiplePriority(column)
              });
              if (originOnClick) {
                originOnClick(event);
              }
            };
            cell.className = (0, _clsx.default)(cell.className, `${prefix}-column-has-sorters`);
            return cell;
          }
        };
      }
      if ('children' in newColumn) {
        newColumn = {
          ...newColumn,
          children: injectSorter(newColumn.children, sorterSates, columnPos)
        };
      }
      return newColumn;
    });
    return injectSorter(innerColumns, mergedSorterStates);
  }, [mergedSorterStates, tableSortDirections, tableShowSorterTooltip, prefix, tableLocale, triggerSorter]);
  const getSorters = (0, _react.useCallback)(() => generateSorterInfo(mergedSorterStates), [mergedSorterStates]);
  return [transformColumns, mergedSorterStates, columnTitleSorterProps, getSorters];
}
