// @universe-design/styles/cjs/ccm-chart-ig-color/index.light.atom-raw.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ccm_chart_L900_raw = exports.ccm_chart_L800_raw = exports.ccm_chart_L700_raw = exports.ccm_chart_L600_raw = exports.ccm_chart_L500_raw = exports.ccm_chart_L50_raw = exports.ccm_chart_L400_raw = exports.ccm_chart_L300_raw = exports.ccm_chart_L200_raw = exports.ccm_chart_L100_raw = exports.ccm_chart_I900_raw = exports.ccm_chart_I800_raw = exports.ccm_chart_I700_raw = exports.ccm_chart_I600_raw = exports.ccm_chart_I500_raw = exports.ccm_chart_I50_raw = exports.ccm_chart_I400_raw = exports.ccm_chart_I300_raw = exports.ccm_chart_I200_raw = exports.ccm_chart_I100_raw = exports.ccm_chart_G900_raw = exports.ccm_chart_G800_raw = exports.ccm_chart_G700_raw = exports.ccm_chart_G600_raw = exports.ccm_chart_G500_raw = exports.ccm_chart_G50_raw = exports.ccm_chart_G400_raw = exports.ccm_chart_G300_raw = exports.ccm_chart_G200_raw = exports.ccm_chart_G100_raw = exports.ccm_chart_C900_raw = exports.ccm_chart_C800_raw = exports.ccm_chart_C700_raw = exports.ccm_chart_C600_raw = exports.ccm_chart_C500_raw = exports.ccm_chart_C50_raw = exports.ccm_chart_C400_raw = exports.ccm_chart_C300_raw = exports.ccm_chart_C200_raw = exports.ccm_chart_C100_raw = exports.ccm_chart_B900_raw = exports.ccm_chart_B800_raw = exports.ccm_chart_B700_raw = exports.ccm_chart_B600_raw = exports.ccm_chart_B500_raw = exports.ccm_chart_B50_raw = exports.ccm_chart_B400_raw = exports.ccm_chart_B300_raw = exports.ccm_chart_B200_raw = exports.ccm_chart_B100_raw = void 0;
exports.ccm_chart_T600_raw = exports.ccm_chart_T500_raw = exports.ccm_chart_T50_raw = exports.ccm_chart_T400_raw = exports.ccm_chart_T300_raw = exports.ccm_chart_T200_raw = exports.ccm_chart_T100_raw = exports.ccm_chart_R900_raw = exports.ccm_chart_R800_raw = exports.ccm_chart_R700_raw = exports.ccm_chart_R600_raw = exports.ccm_chart_R500_raw = exports.ccm_chart_R50_raw = exports.ccm_chart_R400_raw = exports.ccm_chart_R300_raw = exports.ccm_chart_R200_raw = exports.ccm_chart_R100_raw = exports.ccm_chart_P900_raw = exports.ccm_chart_P800_raw = exports.ccm_chart_P700_raw = exports.ccm_chart_P600_raw = exports.ccm_chart_P500_raw = exports.ccm_chart_P50_raw = exports.ccm_chart_P400_raw = exports.ccm_chart_P300_raw = exports.ccm_chart_P200_raw = exports.ccm_chart_P100_raw = exports.ccm_chart_O900_raw = exports.ccm_chart_O800_raw = exports.ccm_chart_O700_raw = exports.ccm_chart_O600_raw = exports.ccm_chart_O500_raw = exports.ccm_chart_O50_raw = exports.ccm_chart_O400_raw = exports.ccm_chart_O300_raw = exports.ccm_chart_O200_raw = exports.ccm_chart_O100_raw = exports.ccm_chart_N900_raw = exports.ccm_chart_N800_raw = exports.ccm_chart_N700_raw = exports.ccm_chart_N650_raw = exports.ccm_chart_N600_raw = exports.ccm_chart_N500_raw = exports.ccm_chart_N50_raw = exports.ccm_chart_N400_raw = exports.ccm_chart_N300_raw = exports.ccm_chart_N200_raw = exports.ccm_chart_N1000_raw = exports.ccm_chart_N100_raw = exports.ccm_chart_N00_raw = void 0;
exports.ccm_chart_pareto_line_default_raw = exports.ccm_chart_pareto_label_default_raw = exports.ccm_chart_Y900_raw = exports.ccm_chart_Y800_raw = exports.ccm_chart_Y700_raw = exports.ccm_chart_Y600_raw = exports.ccm_chart_Y500_raw = exports.ccm_chart_Y50_raw = exports.ccm_chart_Y400_raw = exports.ccm_chart_Y300_raw = exports.ccm_chart_Y200_raw = exports.ccm_chart_Y100_raw = exports.ccm_chart_W900_raw = exports.ccm_chart_W800_raw = exports.ccm_chart_W700_raw = exports.ccm_chart_W600_raw = exports.ccm_chart_W500_raw = exports.ccm_chart_W50_raw = exports.ccm_chart_W400_raw = exports.ccm_chart_W300_raw = exports.ccm_chart_W200_raw = exports.ccm_chart_W100_raw = exports.ccm_chart_V900_raw = exports.ccm_chart_V800_raw = exports.ccm_chart_V700_raw = exports.ccm_chart_V600_raw = exports.ccm_chart_V500_raw = exports.ccm_chart_V50_raw = exports.ccm_chart_V400_raw = exports.ccm_chart_V300_raw = exports.ccm_chart_V200_raw = exports.ccm_chart_V100_raw = exports.ccm_chart_T900_raw = exports.ccm_chart_T800_raw = exports.ccm_chart_T700_raw = void 0;
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
exports.ccm_chart_B100_raw = '225, 234, 255';
exports.ccm_chart_B200_raw = '186, 206, 253';
exports.ccm_chart_B300_raw = '130, 167, 252';
exports.ccm_chart_B400_raw = '78, 131, 253';
exports.ccm_chart_B50_raw = '240, 244, 255';
exports.ccm_chart_B500_raw = '51, 112, 255';
exports.ccm_chart_B600_raw = '36, 91, 219';
exports.ccm_chart_B700_raw = '28, 76, 186';
exports.ccm_chart_B800_raw = '19, 60, 154';
exports.ccm_chart_B900_raw = '12, 41, 110';
exports.ccm_chart_C100_raw = '253, 221, 239';
exports.ccm_chart_C200_raw = '249, 174, 217';
exports.ccm_chart_C300_raw = '245, 122, 192';
exports.ccm_chart_C400_raw = '241, 75, 169';
exports.ccm_chart_C50_raw = '254, 241, 248';
exports.ccm_chart_C500_raw = '240, 29, 148';
exports.ccm_chart_C600_raw = '199, 16, 119';
exports.ccm_chart_C700_raw = '158, 19, 97';
exports.ccm_chart_C800_raw = '122, 15, 75';
exports.ccm_chart_C900_raw = '82, 10, 50';
exports.ccm_chart_G100_raw = '217, 245, 214';
exports.ccm_chart_G200_raw = '183, 237, 177';
exports.ccm_chart_G300_raw = '142, 224, 133';
exports.ccm_chart_G400_raw = '98, 210, 86';
exports.ccm_chart_G50_raw = '240, 251, 239';
exports.ccm_chart_G500_raw = '52, 199, 36';
exports.ccm_chart_G600_raw = '46, 161, 33';
exports.ccm_chart_G700_raw = '35, 123, 25';
exports.ccm_chart_G800_raw = '24, 96, 16';
exports.ccm_chart_G900_raw = '18, 75, 12';
exports.ccm_chart_I100_raw = '224, 226, 250';
exports.ccm_chart_I200_raw = '180, 185, 243';
exports.ccm_chart_I300_raw = '123, 131, 234';
exports.ccm_chart_I400_raw = '97, 106, 229';
exports.ccm_chart_I50_raw = '242, 243, 253';
exports.ccm_chart_I500_raw = '73, 84, 230';
exports.ccm_chart_I600_raw = '41, 51, 199';
exports.ccm_chart_I700_raw = '22, 31, 162';
exports.ccm_chart_I800_raw = '10, 18, 133';
exports.ccm_chart_I900_raw = '12, 18, 100';
exports.ccm_chart_L100_raw = '238, 246, 198';
exports.ccm_chart_L200_raw = '223, 238, 150';
exports.ccm_chart_L300_raw = '210, 231, 106';
exports.ccm_chart_L400_raw = '195, 221, 64';
exports.ccm_chart_L50_raw = '247, 250, 229';
exports.ccm_chart_L500_raw = '179, 214, 0';
exports.ccm_chart_L600_raw = '143, 172, 2';
exports.ccm_chart_L700_raw = '102, 121, 1';
exports.ccm_chart_L800_raw = '73, 87, 0';
exports.ccm_chart_L900_raw = '53, 66, 0';
exports.ccm_chart_N00_raw = '255, 255, 255';
exports.ccm_chart_N100_raw = '242, 243, 245';
exports.ccm_chart_N1000_raw = '0, 0, 0';
exports.ccm_chart_N200_raw = '239, 240, 241';
exports.ccm_chart_N300_raw = '222, 224, 227';
exports.ccm_chart_N400_raw = '187, 191, 196';
exports.ccm_chart_N50_raw = '245, 246, 247';
exports.ccm_chart_N500_raw = '143, 149, 158';
exports.ccm_chart_N600_raw = '100, 106, 115';
exports.ccm_chart_N650_raw = '81, 86, 93';
exports.ccm_chart_N700_raw = '55, 60, 67';
exports.ccm_chart_N800_raw = '43, 47, 54';
exports.ccm_chart_N900_raw = '31, 35, 41';
exports.ccm_chart_O100_raw = '254, 234, 210';
exports.ccm_chart_O200_raw = '254, 212, 164';
exports.ccm_chart_O300_raw = '255, 186, 107';
exports.ccm_chart_O400_raw = '255, 165, 61';
exports.ccm_chart_O50_raw = '255, 245, 235';
exports.ccm_chart_O500_raw = '255, 136, 0';
exports.ccm_chart_O600_raw = '222, 120, 2';
exports.ccm_chart_O700_raw = '178, 98, 6';
exports.ccm_chart_O800_raw = '143, 79, 4';
exports.ccm_chart_O900_raw = '107, 57, 0';
exports.ccm_chart_P100_raw = '236, 226, 254';
exports.ccm_chart_P200_raw = '205, 178, 250';
exports.ccm_chart_P300_raw = '173, 130, 247';
exports.ccm_chart_P400_raw = '147, 90, 246';
exports.ccm_chart_P50_raw = '246, 241, 254';
exports.ccm_chart_P500_raw = '127, 59, 245';
exports.ccm_chart_P600_raw = '100, 37, 208';
exports.ccm_chart_P700_raw = '78, 27, 167';
exports.ccm_chart_P800_raw = '56, 13, 130';
exports.ccm_chart_P900_raw = '39, 5, 97';
exports.ccm_chart_R100_raw = '253, 226, 226';
exports.ccm_chart_R200_raw = '251, 191, 188';
exports.ccm_chart_R300_raw = '249, 142, 139';
exports.ccm_chart_R400_raw = '247, 105, 100';
exports.ccm_chart_R50_raw = '254, 241, 241';
exports.ccm_chart_R500_raw = '245, 74, 69';
exports.ccm_chart_R600_raw = '216, 57, 49';
exports.ccm_chart_R700_raw = '172, 47, 40';
exports.ccm_chart_R800_raw = '129, 37, 32';
exports.ccm_chart_R900_raw = '98, 28, 24';
exports.ccm_chart_T100_raw = '213, 246, 242';
exports.ccm_chart_T200_raw = '169, 239, 230';
exports.ccm_chart_T300_raw = '100, 232, 214';
exports.ccm_chart_T400_raw = '20, 225, 198';
exports.ccm_chart_T50_raw = '234, 250, 248';
exports.ccm_chart_T500_raw = '0, 214, 185';
exports.ccm_chart_T600_raw = '4, 180, 156';
exports.ccm_chart_T700_raw = '7, 131, 114';
exports.ccm_chart_T800_raw = '3, 99, 86';
exports.ccm_chart_T900_raw = '2, 75, 65';
exports.ccm_chart_V100_raw = '248, 222, 248';
exports.ccm_chart_V200_raw = '239, 185, 239';
exports.ccm_chart_V300_raw = '229, 143, 229';
exports.ccm_chart_V400_raw = '219, 102, 219';
exports.ccm_chart_V50_raw = '252, 238, 252';
exports.ccm_chart_V500_raw = '209, 54, 209';
exports.ccm_chart_V600_raw = '178, 32, 178';
exports.ccm_chart_V700_raw = '140, 33, 140';
exports.ccm_chart_V800_raw = '104, 18, 104';
exports.ccm_chart_V900_raw = '70, 11, 70';
exports.ccm_chart_W100_raw = '217, 243, 253';
exports.ccm_chart_W200_raw = '177, 232, 252';
exports.ccm_chart_W300_raw = '126, 218, 251';
exports.ccm_chart_W400_raw = '80, 206, 251';
exports.ccm_chart_W50_raw = '236, 249, 254';
exports.ccm_chart_W500_raw = '20, 192, 255';
exports.ccm_chart_W600_raw = '4, 159, 215';
exports.ccm_chart_W700_raw = '3, 126, 170';
exports.ccm_chart_W800_raw = '0, 97, 133';
exports.ccm_chart_W900_raw = '0, 71, 97';
exports.ccm_chart_Y100_raw = '250, 241, 209';
exports.ccm_chart_Y200_raw = '248, 230, 171';
exports.ccm_chart_Y300_raw = '247, 220, 130';
exports.ccm_chart_Y400_raw = '250, 211, 85';
exports.ccm_chart_Y50_raw = '253, 249, 237';
exports.ccm_chart_Y500_raw = '255, 198, 10';
exports.ccm_chart_Y600_raw = '220, 155, 4';
exports.ccm_chart_Y700_raw = '170, 120, 3';
exports.ccm_chart_Y800_raw = '121, 81, 1';
exports.ccm_chart_Y900_raw = '92, 58, 0';
exports.ccm_chart_pareto_label_default_raw = '31, 35, 41';
exports.ccm_chart_pareto_line_default_raw = '229, 206, 0';
