// @universe-design/styles/cjs/ccm-chart-ig-color/index.dark.atom-raw.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ccm_chart_L900_raw = exports.ccm_chart_L800_raw = exports.ccm_chart_L700_raw = exports.ccm_chart_L600_raw = exports.ccm_chart_L500_raw = exports.ccm_chart_L50_raw = exports.ccm_chart_L400_raw = exports.ccm_chart_L300_raw = exports.ccm_chart_L200_raw = exports.ccm_chart_L100_raw = exports.ccm_chart_I900_raw = exports.ccm_chart_I800_raw = exports.ccm_chart_I700_raw = exports.ccm_chart_I600_raw = exports.ccm_chart_I500_raw = exports.ccm_chart_I50_raw = exports.ccm_chart_I400_raw = exports.ccm_chart_I300_raw = exports.ccm_chart_I200_raw = exports.ccm_chart_I100_raw = exports.ccm_chart_G900_raw = exports.ccm_chart_G800_raw = exports.ccm_chart_G700_raw = exports.ccm_chart_G600_raw = exports.ccm_chart_G500_raw = exports.ccm_chart_G50_raw = exports.ccm_chart_G400_raw = exports.ccm_chart_G300_raw = exports.ccm_chart_G200_raw = exports.ccm_chart_G100_raw = exports.ccm_chart_C900_raw = exports.ccm_chart_C800_raw = exports.ccm_chart_C700_raw = exports.ccm_chart_C600_raw = exports.ccm_chart_C500_raw = exports.ccm_chart_C50_raw = exports.ccm_chart_C400_raw = exports.ccm_chart_C300_raw = exports.ccm_chart_C200_raw = exports.ccm_chart_C100_raw = exports.ccm_chart_B900_raw = exports.ccm_chart_B800_raw = exports.ccm_chart_B700_raw = exports.ccm_chart_B600_raw = exports.ccm_chart_B500_raw = exports.ccm_chart_B50_raw = exports.ccm_chart_B400_raw = exports.ccm_chart_B300_raw = exports.ccm_chart_B200_raw = exports.ccm_chart_B100_raw = void 0;
exports.ccm_chart_T600_raw = exports.ccm_chart_T500_raw = exports.ccm_chart_T50_raw = exports.ccm_chart_T400_raw = exports.ccm_chart_T300_raw = exports.ccm_chart_T200_raw = exports.ccm_chart_T100_raw = exports.ccm_chart_R900_raw = exports.ccm_chart_R800_raw = exports.ccm_chart_R700_raw = exports.ccm_chart_R600_raw = exports.ccm_chart_R500_raw = exports.ccm_chart_R50_raw = exports.ccm_chart_R400_raw = exports.ccm_chart_R300_raw = exports.ccm_chart_R200_raw = exports.ccm_chart_R100_raw = exports.ccm_chart_P900_raw = exports.ccm_chart_P800_raw = exports.ccm_chart_P700_raw = exports.ccm_chart_P600_raw = exports.ccm_chart_P500_raw = exports.ccm_chart_P50_raw = exports.ccm_chart_P400_raw = exports.ccm_chart_P300_raw = exports.ccm_chart_P200_raw = exports.ccm_chart_P100_raw = exports.ccm_chart_O900_raw = exports.ccm_chart_O800_raw = exports.ccm_chart_O700_raw = exports.ccm_chart_O600_raw = exports.ccm_chart_O500_raw = exports.ccm_chart_O50_raw = exports.ccm_chart_O400_raw = exports.ccm_chart_O300_raw = exports.ccm_chart_O200_raw = exports.ccm_chart_O100_raw = exports.ccm_chart_N900_raw = exports.ccm_chart_N800_raw = exports.ccm_chart_N700_raw = exports.ccm_chart_N650_raw = exports.ccm_chart_N600_raw = exports.ccm_chart_N500_raw = exports.ccm_chart_N50_raw = exports.ccm_chart_N400_raw = exports.ccm_chart_N300_raw = exports.ccm_chart_N200_raw = exports.ccm_chart_N1000_raw = exports.ccm_chart_N100_raw = exports.ccm_chart_N00_raw = void 0;
exports.ccm_chart_pareto_line_default_raw = exports.ccm_chart_pareto_label_default_raw = exports.ccm_chart_Y900_raw = exports.ccm_chart_Y800_raw = exports.ccm_chart_Y700_raw = exports.ccm_chart_Y600_raw = exports.ccm_chart_Y500_raw = exports.ccm_chart_Y50_raw = exports.ccm_chart_Y400_raw = exports.ccm_chart_Y300_raw = exports.ccm_chart_Y200_raw = exports.ccm_chart_Y100_raw = exports.ccm_chart_W900_raw = exports.ccm_chart_W800_raw = exports.ccm_chart_W700_raw = exports.ccm_chart_W600_raw = exports.ccm_chart_W500_raw = exports.ccm_chart_W50_raw = exports.ccm_chart_W400_raw = exports.ccm_chart_W300_raw = exports.ccm_chart_W200_raw = exports.ccm_chart_W100_raw = exports.ccm_chart_V900_raw = exports.ccm_chart_V800_raw = exports.ccm_chart_V700_raw = exports.ccm_chart_V600_raw = exports.ccm_chart_V500_raw = exports.ccm_chart_V50_raw = exports.ccm_chart_V400_raw = exports.ccm_chart_V300_raw = exports.ccm_chart_V200_raw = exports.ccm_chart_V100_raw = exports.ccm_chart_T900_raw = exports.ccm_chart_T800_raw = exports.ccm_chart_T700_raw = void 0;
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
exports.ccm_chart_B100_raw = '25, 42, 76';
exports.ccm_chart_B200_raw = '32, 62, 120';
exports.ccm_chart_B300_raw = '40, 81, 163';
exports.ccm_chart_B400_raw = '46, 101, 209';
exports.ccm_chart_B50_raw = '21, 31, 51';
exports.ccm_chart_B500_raw = '76, 136, 255';
exports.ccm_chart_B600_raw = '112, 160, 255';
exports.ccm_chart_B700_raw = '153, 187, 255';
exports.ccm_chart_B800_raw = '184, 207, 255';
exports.ccm_chart_B900_raw = '206, 223, 255';
exports.ccm_chart_C100_raw = '92, 37, 66';
exports.ccm_chart_C200_raw = '145, 55, 106';
exports.ccm_chart_C300_raw = '173, 64, 126';
exports.ccm_chart_C400_raw = '204, 71, 146';
exports.ccm_chart_C50_raw = '66, 27, 49';
exports.ccm_chart_C500_raw = '227, 82, 163';
exports.ccm_chart_C600_raw = '237, 102, 179';
exports.ccm_chart_C700_raw = '255, 145, 207';
exports.ccm_chart_C800_raw = '255, 189, 226';
exports.ccm_chart_C900_raw = '255, 219, 239';
exports.ccm_chart_G100_raw = '31, 71, 27';
exports.ccm_chart_G200_raw = '41, 107, 34';
exports.ccm_chart_G300_raw = '51, 148, 40';
exports.ccm_chart_G400_raw = '59, 176, 46';
exports.ccm_chart_G50_raw = '25, 48, 22';
exports.ccm_chart_G500_raw = '84, 194, 72';
exports.ccm_chart_G600_raw = '109, 209, 98';
exports.ccm_chart_G700_raw = '146, 222, 137';
exports.ccm_chart_G800_raw = '189, 240, 184';
exports.ccm_chart_G900_raw = '214, 245, 210';
exports.ccm_chart_I100_raw = '38, 42, 94';
exports.ccm_chart_I200_raw = '53, 59, 140';
exports.ccm_chart_I300_raw = '65, 73, 181';
exports.ccm_chart_I400_raw = '76, 85, 212';
exports.ccm_chart_I50_raw = '29, 31, 71';
exports.ccm_chart_I500_raw = '95, 105, 237';
exports.ccm_chart_I600_raw = '115, 123, 235';
exports.ccm_chart_I700_raw = '158, 165, 255';
exports.ccm_chart_I800_raw = '187, 192, 246';
exports.ccm_chart_I900_raw = '210, 213, 253';
exports.ccm_chart_L100_raw = '73, 87, 6';
exports.ccm_chart_L200_raw = '109, 130, 5';
exports.ccm_chart_L300_raw = '145, 173, 2';
exports.ccm_chart_L400_raw = '171, 201, 18';
exports.ccm_chart_L50_raw = '46, 54, 5';
exports.ccm_chart_L500_raw = '185, 214, 39';
exports.ccm_chart_L600_raw = '199, 224, 77';
exports.ccm_chart_L700_raw = '214, 234, 117';
exports.ccm_chart_L800_raw = '231, 245, 166';
exports.ccm_chart_L900_raw = '246, 255, 204';
exports.ccm_chart_N00_raw = '10, 10, 10';
exports.ccm_chart_N100_raw = '41, 41, 41';
exports.ccm_chart_N1000_raw = '255, 255, 255';
exports.ccm_chart_N200_raw = '55, 55, 55';
exports.ccm_chart_N300_raw = '67, 67, 67';
exports.ccm_chart_N400_raw = '95, 95, 95';
exports.ccm_chart_N50_raw = '26, 26, 26';
exports.ccm_chart_N500_raw = '117, 117, 117';
exports.ccm_chart_N600_raw = '166, 166, 166';
exports.ccm_chart_N650_raw = '207, 207, 207';
exports.ccm_chart_N700_raw = '224, 224, 224';
exports.ccm_chart_N800_raw = '232, 232, 232';
exports.ccm_chart_N900_raw = '235, 235, 235';
exports.ccm_chart_O100_raw = '87, 51, 10';
exports.ccm_chart_O200_raw = '132, 81, 23';
exports.ccm_chart_O300_raw = '189, 112, 23';
exports.ccm_chart_O400_raw = '222, 130, 24';
exports.ccm_chart_O50_raw = '51, 33, 11';
exports.ccm_chart_O500_raw = '242, 150, 44';
exports.ccm_chart_O600_raw = '245, 165, 74';
exports.ccm_chart_O700_raw = '250, 188, 117';
exports.ccm_chart_O800_raw = '255, 216, 172';
exports.ccm_chart_O900_raw = '255, 234, 206';
exports.ccm_chart_P100_raw = '54, 29, 97';
exports.ccm_chart_P200_raw = '77, 38, 145';
exports.ccm_chart_P300_raw = '101, 51, 189';
exports.ccm_chart_P400_raw = '124, 74, 212';
exports.ccm_chart_P50_raw = '40, 23, 69';
exports.ccm_chart_P500_raw = '151, 98, 245';
exports.ccm_chart_P600_raw = '164, 114, 252';
exports.ccm_chart_P700_raw = '185, 144, 255';
exports.ccm_chart_P800_raw = '211, 186, 253';
exports.ccm_chart_P900_raw = '227, 212, 255';
exports.ccm_chart_R100_raw = '74, 29, 27';
exports.ccm_chart_R200_raw = '107, 40, 39';
exports.ccm_chart_R300_raw = '158, 56, 54';
exports.ccm_chart_R400_raw = '204, 71, 67';
exports.ccm_chart_R50_raw = '54, 25, 24';
exports.ccm_chart_R500_raw = '240, 91, 86';
exports.ccm_chart_R600_raw = '250, 120, 115';
exports.ccm_chart_R700_raw = '255, 156, 153';
exports.ccm_chart_R800_raw = '254, 198, 195';
exports.ccm_chart_R900_raw = '255, 224, 224';
exports.ccm_chart_T100_raw = '21, 61, 56';
exports.ccm_chart_T200_raw = '35, 102, 93';
exports.ccm_chart_T300_raw = '38, 150, 135';
exports.ccm_chart_T400_raw = '55, 184, 166';
exports.ccm_chart_T50_raw = '22, 46, 43';
exports.ccm_chart_T500_raw = '65, 204, 185';
exports.ccm_chart_T600_raw = '92, 219, 202';
exports.ccm_chart_T700_raw = '127, 235, 220';
exports.ccm_chart_T800_raw = '183, 247, 239';
exports.ccm_chart_T900_raw = '218, 249, 245';
exports.ccm_chart_V100_raw = '79, 28, 79';
exports.ccm_chart_V200_raw = '122, 40, 122';
exports.ccm_chart_V300_raw = '148, 59, 148';
exports.ccm_chart_V400_raw = '181, 74, 181';
exports.ccm_chart_V50_raw = '59, 21, 59';
exports.ccm_chart_V500_raw = '212, 87, 212';
exports.ccm_chart_V600_raw = '222, 115, 222';
exports.ccm_chart_V700_raw = '232, 152, 232';
exports.ccm_chart_V800_raw = '242, 191, 242';
exports.ccm_chart_V900_raw = '251, 218, 251';
exports.ccm_chart_W100_raw = '23, 55, 66';
exports.ccm_chart_W200_raw = '34, 91, 112';
exports.ccm_chart_W300_raw = '39, 130, 163';
exports.ccm_chart_W400_raw = '46, 165, 209';
exports.ccm_chart_W50_raw = '20, 41, 48';
exports.ccm_chart_W500_raw = '66, 189, 235';
exports.ccm_chart_W600_raw = '95, 212, 254';
exports.ccm_chart_W700_raw = '137, 223, 254';
exports.ccm_chart_W800_raw = '194, 238, 255';
exports.ccm_chart_W900_raw = '222, 246, 255';
exports.ccm_chart_Y100_raw = '87, 71, 17';
exports.ccm_chart_Y200_raw = '143, 114, 17';
exports.ccm_chart_Y300_raw = '201, 157, 8';
exports.ccm_chart_Y400_raw = '240, 187, 10';
exports.ccm_chart_Y50_raw = '56, 46, 12';
exports.ccm_chart_Y500_raw = '250, 200, 35';
exports.ccm_chart_Y600_raw = '252, 212, 86';
exports.ccm_chart_Y700_raw = '255, 222, 117';
exports.ccm_chart_Y800_raw = '255, 230, 148';
exports.ccm_chart_Y900_raw = '252, 239, 194';
exports.ccm_chart_pareto_label_default_raw = '235, 235, 235';
exports.ccm_chart_pareto_line_default_raw = '250, 237, 87';
