// @universe-design/styles/cjs/ccm-chart-ig-color/index.light.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ccm_chart_L900 = exports.ccm_chart_L800 = exports.ccm_chart_L700 = exports.ccm_chart_L600 = exports.ccm_chart_L500 = exports.ccm_chart_L50 = exports.ccm_chart_L400 = exports.ccm_chart_L300 = exports.ccm_chart_L200 = exports.ccm_chart_L100 = exports.ccm_chart_I900 = exports.ccm_chart_I800 = exports.ccm_chart_I700 = exports.ccm_chart_I600 = exports.ccm_chart_I500 = exports.ccm_chart_I50 = exports.ccm_chart_I400 = exports.ccm_chart_I300 = exports.ccm_chart_I200 = exports.ccm_chart_I100 = exports.ccm_chart_G900 = exports.ccm_chart_G800 = exports.ccm_chart_G700 = exports.ccm_chart_G600 = exports.ccm_chart_G500 = exports.ccm_chart_G50 = exports.ccm_chart_G400 = exports.ccm_chart_G300 = exports.ccm_chart_G200 = exports.ccm_chart_G100 = exports.ccm_chart_C900 = exports.ccm_chart_C800 = exports.ccm_chart_C700 = exports.ccm_chart_C600 = exports.ccm_chart_C500 = exports.ccm_chart_C50 = exports.ccm_chart_C400 = exports.ccm_chart_C300 = exports.ccm_chart_C200 = exports.ccm_chart_C100 = exports.ccm_chart_B900 = exports.ccm_chart_B800 = exports.ccm_chart_B700 = exports.ccm_chart_B600 = exports.ccm_chart_B500 = exports.ccm_chart_B50 = exports.ccm_chart_B400 = exports.ccm_chart_B300 = exports.ccm_chart_B200 = exports.ccm_chart_B100 = void 0;
exports.ccm_chart_T600 = exports.ccm_chart_T500 = exports.ccm_chart_T50 = exports.ccm_chart_T400 = exports.ccm_chart_T300 = exports.ccm_chart_T200 = exports.ccm_chart_T100 = exports.ccm_chart_R900 = exports.ccm_chart_R800 = exports.ccm_chart_R700 = exports.ccm_chart_R600 = exports.ccm_chart_R500 = exports.ccm_chart_R50 = exports.ccm_chart_R400 = exports.ccm_chart_R300 = exports.ccm_chart_R200 = exports.ccm_chart_R100 = exports.ccm_chart_P900 = exports.ccm_chart_P800 = exports.ccm_chart_P700 = exports.ccm_chart_P600 = exports.ccm_chart_P500 = exports.ccm_chart_P50 = exports.ccm_chart_P400 = exports.ccm_chart_P300 = exports.ccm_chart_P200 = exports.ccm_chart_P100 = exports.ccm_chart_O900 = exports.ccm_chart_O800 = exports.ccm_chart_O700 = exports.ccm_chart_O600 = exports.ccm_chart_O500 = exports.ccm_chart_O50 = exports.ccm_chart_O400 = exports.ccm_chart_O300 = exports.ccm_chart_O200 = exports.ccm_chart_O100 = exports.ccm_chart_N900 = exports.ccm_chart_N800 = exports.ccm_chart_N700 = exports.ccm_chart_N650 = exports.ccm_chart_N600 = exports.ccm_chart_N500 = exports.ccm_chart_N50 = exports.ccm_chart_N400 = exports.ccm_chart_N300 = exports.ccm_chart_N200 = exports.ccm_chart_N1000 = exports.ccm_chart_N100 = exports.ccm_chart_N00 = void 0;
exports.ccm_chart_pareto_line_default = exports.ccm_chart_pareto_label_default = exports.ccm_chart_Y900 = exports.ccm_chart_Y800 = exports.ccm_chart_Y700 = exports.ccm_chart_Y600 = exports.ccm_chart_Y500 = exports.ccm_chart_Y50 = exports.ccm_chart_Y400 = exports.ccm_chart_Y300 = exports.ccm_chart_Y200 = exports.ccm_chart_Y100 = exports.ccm_chart_W900 = exports.ccm_chart_W800 = exports.ccm_chart_W700 = exports.ccm_chart_W600 = exports.ccm_chart_W500 = exports.ccm_chart_W50 = exports.ccm_chart_W400 = exports.ccm_chart_W300 = exports.ccm_chart_W200 = exports.ccm_chart_W100 = exports.ccm_chart_V900 = exports.ccm_chart_V800 = exports.ccm_chart_V700 = exports.ccm_chart_V600 = exports.ccm_chart_V500 = exports.ccm_chart_V50 = exports.ccm_chart_V400 = exports.ccm_chart_V300 = exports.ccm_chart_V200 = exports.ccm_chart_V100 = exports.ccm_chart_T900 = exports.ccm_chart_T800 = exports.ccm_chart_T700 = void 0;
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
exports.ccm_chart_B100 = '#e1eaff';
exports.ccm_chart_B200 = '#bacefd';
exports.ccm_chart_B300 = '#82a7fc';
exports.ccm_chart_B400 = '#4e83fd';
exports.ccm_chart_B50 = '#f0f4ff';
exports.ccm_chart_B500 = '#3370ff';
exports.ccm_chart_B600 = '#245bdb';
exports.ccm_chart_B700 = '#1c4cba';
exports.ccm_chart_B800 = '#133c9a';
exports.ccm_chart_B900 = '#0c296e';
exports.ccm_chart_C100 = '#fdddef';
exports.ccm_chart_C200 = '#f9aed9';
exports.ccm_chart_C300 = '#f57ac0';
exports.ccm_chart_C400 = '#f14ba9';
exports.ccm_chart_C50 = '#fef1f8';
exports.ccm_chart_C500 = '#f01d94';
exports.ccm_chart_C600 = '#c71077';
exports.ccm_chart_C700 = '#9e1361';
exports.ccm_chart_C800 = '#7a0f4b';
exports.ccm_chart_C900 = '#520a32';
exports.ccm_chart_G100 = '#d9f5d6';
exports.ccm_chart_G200 = '#b7edb1';
exports.ccm_chart_G300 = '#8ee085';
exports.ccm_chart_G400 = '#62d256';
exports.ccm_chart_G50 = '#f0fbef';
exports.ccm_chart_G500 = '#34c724';
exports.ccm_chart_G600 = '#2ea121';
exports.ccm_chart_G700 = '#237b19';
exports.ccm_chart_G800 = '#186010';
exports.ccm_chart_G900 = '#124b0c';
exports.ccm_chart_I100 = '#e0e2fa';
exports.ccm_chart_I200 = '#b4b9f3';
exports.ccm_chart_I300 = '#7b83ea';
exports.ccm_chart_I400 = '#616ae5';
exports.ccm_chart_I50 = '#f2f3fd';
exports.ccm_chart_I500 = '#4954e6';
exports.ccm_chart_I600 = '#2933c7';
exports.ccm_chart_I700 = '#161fa2';
exports.ccm_chart_I800 = '#0a1285';
exports.ccm_chart_I900 = '#0c1264';
exports.ccm_chart_L100 = '#eef6c6';
exports.ccm_chart_L200 = '#dfee96';
exports.ccm_chart_L300 = '#d2e76a';
exports.ccm_chart_L400 = '#c3dd40';
exports.ccm_chart_L50 = '#f7fae5';
exports.ccm_chart_L500 = '#b3d600';
exports.ccm_chart_L600 = '#8fac02';
exports.ccm_chart_L700 = '#667901';
exports.ccm_chart_L800 = '#495700';
exports.ccm_chart_L900 = '#354200';
exports.ccm_chart_N00 = '#ffffff';
exports.ccm_chart_N100 = '#f2f3f5';
exports.ccm_chart_N1000 = '#000000';
exports.ccm_chart_N200 = '#eff0f1';
exports.ccm_chart_N300 = '#dee0e3';
exports.ccm_chart_N400 = '#bbbfc4';
exports.ccm_chart_N50 = '#f5f6f7';
exports.ccm_chart_N500 = '#8f959e';
exports.ccm_chart_N600 = '#646a73';
exports.ccm_chart_N650 = '#51565d';
exports.ccm_chart_N700 = '#373c43';
exports.ccm_chart_N800 = '#2b2f36';
exports.ccm_chart_N900 = '#1f2329';
exports.ccm_chart_O100 = '#feead2';
exports.ccm_chart_O200 = '#fed4a4';
exports.ccm_chart_O300 = '#ffba6b';
exports.ccm_chart_O400 = '#ffa53d';
exports.ccm_chart_O50 = '#fff5eb';
exports.ccm_chart_O500 = '#ff8800';
exports.ccm_chart_O600 = '#de7802';
exports.ccm_chart_O700 = '#b26206';
exports.ccm_chart_O800 = '#8f4f04';
exports.ccm_chart_O900 = '#6b3900';
exports.ccm_chart_P100 = '#ece2fe';
exports.ccm_chart_P200 = '#cdb2fa';
exports.ccm_chart_P300 = '#ad82f7';
exports.ccm_chart_P400 = '#935af6';
exports.ccm_chart_P50 = '#f6f1fe';
exports.ccm_chart_P500 = '#7f3bf5';
exports.ccm_chart_P600 = '#6425d0';
exports.ccm_chart_P700 = '#4e1ba7';
exports.ccm_chart_P800 = '#380d82';
exports.ccm_chart_P900 = '#270561';
exports.ccm_chart_R100 = '#fde2e2';
exports.ccm_chart_R200 = '#fbbfbc';
exports.ccm_chart_R300 = '#f98e8b';
exports.ccm_chart_R400 = '#f76964';
exports.ccm_chart_R50 = '#fef1f1';
exports.ccm_chart_R500 = '#f54a45';
exports.ccm_chart_R600 = '#d83931';
exports.ccm_chart_R700 = '#ac2f28';
exports.ccm_chart_R800 = '#812520';
exports.ccm_chart_R900 = '#621c18';
exports.ccm_chart_T100 = '#d5f6f2';
exports.ccm_chart_T200 = '#a9efe6';
exports.ccm_chart_T300 = '#64e8d6';
exports.ccm_chart_T400 = '#14e1c6';
exports.ccm_chart_T50 = '#eafaf8';
exports.ccm_chart_T500 = '#00d6b9';
exports.ccm_chart_T600 = '#04b49c';
exports.ccm_chart_T700 = '#078372';
exports.ccm_chart_T800 = '#036356';
exports.ccm_chart_T900 = '#024b41';
exports.ccm_chart_V100 = '#f8def8';
exports.ccm_chart_V200 = '#efb9ef';
exports.ccm_chart_V300 = '#e58fe5';
exports.ccm_chart_V400 = '#db66db';
exports.ccm_chart_V50 = '#fceefc';
exports.ccm_chart_V500 = '#d136d1';
exports.ccm_chart_V600 = '#b220b2';
exports.ccm_chart_V700 = '#8c218c';
exports.ccm_chart_V800 = '#681268';
exports.ccm_chart_V900 = '#460b46';
exports.ccm_chart_W100 = '#d9f3fd';
exports.ccm_chart_W200 = '#b1e8fc';
exports.ccm_chart_W300 = '#7edafb';
exports.ccm_chart_W400 = '#50cefb';
exports.ccm_chart_W50 = '#ecf9fe';
exports.ccm_chart_W500 = '#14c0ff';
exports.ccm_chart_W600 = '#049fd7';
exports.ccm_chart_W700 = '#037eaa';
exports.ccm_chart_W800 = '#006185';
exports.ccm_chart_W900 = '#004761';
exports.ccm_chart_Y100 = '#faf1d1';
exports.ccm_chart_Y200 = '#f8e6ab';
exports.ccm_chart_Y300 = '#f7dc82';
exports.ccm_chart_Y400 = '#fad355';
exports.ccm_chart_Y50 = '#fdf9ed';
exports.ccm_chart_Y500 = '#ffc60a';
exports.ccm_chart_Y600 = '#dc9b04';
exports.ccm_chart_Y700 = '#aa7803';
exports.ccm_chart_Y800 = '#795101';
exports.ccm_chart_Y900 = '#5c3a00';
exports.ccm_chart_pareto_label_default = '#1f2329';
exports.ccm_chart_pareto_line_default = '#e5ce00';
