// @universe-design/styles/cjs/ccm-chart-ig-color/index.dark.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ccm_chart_L900 = exports.ccm_chart_L800 = exports.ccm_chart_L700 = exports.ccm_chart_L600 = exports.ccm_chart_L500 = exports.ccm_chart_L50 = exports.ccm_chart_L400 = exports.ccm_chart_L300 = exports.ccm_chart_L200 = exports.ccm_chart_L100 = exports.ccm_chart_I900 = exports.ccm_chart_I800 = exports.ccm_chart_I700 = exports.ccm_chart_I600 = exports.ccm_chart_I500 = exports.ccm_chart_I50 = exports.ccm_chart_I400 = exports.ccm_chart_I300 = exports.ccm_chart_I200 = exports.ccm_chart_I100 = exports.ccm_chart_G900 = exports.ccm_chart_G800 = exports.ccm_chart_G700 = exports.ccm_chart_G600 = exports.ccm_chart_G500 = exports.ccm_chart_G50 = exports.ccm_chart_G400 = exports.ccm_chart_G300 = exports.ccm_chart_G200 = exports.ccm_chart_G100 = exports.ccm_chart_C900 = exports.ccm_chart_C800 = exports.ccm_chart_C700 = exports.ccm_chart_C600 = exports.ccm_chart_C500 = exports.ccm_chart_C50 = exports.ccm_chart_C400 = exports.ccm_chart_C300 = exports.ccm_chart_C200 = exports.ccm_chart_C100 = exports.ccm_chart_B900 = exports.ccm_chart_B800 = exports.ccm_chart_B700 = exports.ccm_chart_B600 = exports.ccm_chart_B500 = exports.ccm_chart_B50 = exports.ccm_chart_B400 = exports.ccm_chart_B300 = exports.ccm_chart_B200 = exports.ccm_chart_B100 = void 0;
exports.ccm_chart_T600 = exports.ccm_chart_T500 = exports.ccm_chart_T50 = exports.ccm_chart_T400 = exports.ccm_chart_T300 = exports.ccm_chart_T200 = exports.ccm_chart_T100 = exports.ccm_chart_R900 = exports.ccm_chart_R800 = exports.ccm_chart_R700 = exports.ccm_chart_R600 = exports.ccm_chart_R500 = exports.ccm_chart_R50 = exports.ccm_chart_R400 = exports.ccm_chart_R300 = exports.ccm_chart_R200 = exports.ccm_chart_R100 = exports.ccm_chart_P900 = exports.ccm_chart_P800 = exports.ccm_chart_P700 = exports.ccm_chart_P600 = exports.ccm_chart_P500 = exports.ccm_chart_P50 = exports.ccm_chart_P400 = exports.ccm_chart_P300 = exports.ccm_chart_P200 = exports.ccm_chart_P100 = exports.ccm_chart_O900 = exports.ccm_chart_O800 = exports.ccm_chart_O700 = exports.ccm_chart_O600 = exports.ccm_chart_O500 = exports.ccm_chart_O50 = exports.ccm_chart_O400 = exports.ccm_chart_O300 = exports.ccm_chart_O200 = exports.ccm_chart_O100 = exports.ccm_chart_N900 = exports.ccm_chart_N800 = exports.ccm_chart_N700 = exports.ccm_chart_N650 = exports.ccm_chart_N600 = exports.ccm_chart_N500 = exports.ccm_chart_N50 = exports.ccm_chart_N400 = exports.ccm_chart_N300 = exports.ccm_chart_N200 = exports.ccm_chart_N1000 = exports.ccm_chart_N100 = exports.ccm_chart_N00 = void 0;
exports.ccm_chart_pareto_line_default = exports.ccm_chart_pareto_label_default = exports.ccm_chart_Y900 = exports.ccm_chart_Y800 = exports.ccm_chart_Y700 = exports.ccm_chart_Y600 = exports.ccm_chart_Y500 = exports.ccm_chart_Y50 = exports.ccm_chart_Y400 = exports.ccm_chart_Y300 = exports.ccm_chart_Y200 = exports.ccm_chart_Y100 = exports.ccm_chart_W900 = exports.ccm_chart_W800 = exports.ccm_chart_W700 = exports.ccm_chart_W600 = exports.ccm_chart_W500 = exports.ccm_chart_W50 = exports.ccm_chart_W400 = exports.ccm_chart_W300 = exports.ccm_chart_W200 = exports.ccm_chart_W100 = exports.ccm_chart_V900 = exports.ccm_chart_V800 = exports.ccm_chart_V700 = exports.ccm_chart_V600 = exports.ccm_chart_V500 = exports.ccm_chart_V50 = exports.ccm_chart_V400 = exports.ccm_chart_V300 = exports.ccm_chart_V200 = exports.ccm_chart_V100 = exports.ccm_chart_T900 = exports.ccm_chart_T800 = exports.ccm_chart_T700 = void 0;
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
exports.ccm_chart_B100 = '#192a4c';
exports.ccm_chart_B200 = '#203e78';
exports.ccm_chart_B300 = '#2851a3';
exports.ccm_chart_B400 = '#2e65d1';
exports.ccm_chart_B50 = '#151f33';
exports.ccm_chart_B500 = '#4c88ff';
exports.ccm_chart_B600 = '#70a0ff';
exports.ccm_chart_B700 = '#99bbff';
exports.ccm_chart_B800 = '#b8cfff';
exports.ccm_chart_B900 = '#cedfff';
exports.ccm_chart_C100 = '#5c2542';
exports.ccm_chart_C200 = '#91376a';
exports.ccm_chart_C300 = '#ad407e';
exports.ccm_chart_C400 = '#cc4792';
exports.ccm_chart_C50 = '#421b31';
exports.ccm_chart_C500 = '#e352a3';
exports.ccm_chart_C600 = '#ed66b3';
exports.ccm_chart_C700 = '#ff91cf';
exports.ccm_chart_C800 = '#ffbde2';
exports.ccm_chart_C900 = '#ffdbef';
exports.ccm_chart_G100 = '#1f471b';
exports.ccm_chart_G200 = '#296b22';
exports.ccm_chart_G300 = '#339428';
exports.ccm_chart_G400 = '#3bb02e';
exports.ccm_chart_G50 = '#193016';
exports.ccm_chart_G500 = '#54c248';
exports.ccm_chart_G600 = '#6dd162';
exports.ccm_chart_G700 = '#92de89';
exports.ccm_chart_G800 = '#bdf0b8';
exports.ccm_chart_G900 = '#d6f5d2';
exports.ccm_chart_I100 = '#262a5e';
exports.ccm_chart_I200 = '#353b8c';
exports.ccm_chart_I300 = '#4149b5';
exports.ccm_chart_I400 = '#4c55d4';
exports.ccm_chart_I50 = '#1d1f47';
exports.ccm_chart_I500 = '#5f69ed';
exports.ccm_chart_I600 = '#737beb';
exports.ccm_chart_I700 = '#9ea5ff';
exports.ccm_chart_I800 = '#bbc0f6';
exports.ccm_chart_I900 = '#d2d5fd';
exports.ccm_chart_L100 = '#495706';
exports.ccm_chart_L200 = '#6d8205';
exports.ccm_chart_L300 = '#91ad02';
exports.ccm_chart_L400 = '#abc912';
exports.ccm_chart_L50 = '#2e3605';
exports.ccm_chart_L500 = '#b9d627';
exports.ccm_chart_L600 = '#c7e04d';
exports.ccm_chart_L700 = '#d6ea75';
exports.ccm_chart_L800 = '#e7f5a6';
exports.ccm_chart_L900 = '#f6ffcc';
exports.ccm_chart_N00 = '#0a0a0a';
exports.ccm_chart_N100 = '#292929';
exports.ccm_chart_N1000 = '#ffffff';
exports.ccm_chart_N200 = '#373737';
exports.ccm_chart_N300 = '#434343';
exports.ccm_chart_N400 = '#5f5f5f';
exports.ccm_chart_N50 = '#1a1a1a';
exports.ccm_chart_N500 = '#757575';
exports.ccm_chart_N600 = '#a6a6a6';
exports.ccm_chart_N650 = '#cfcfcf';
exports.ccm_chart_N700 = '#e0e0e0';
exports.ccm_chart_N800 = '#e8e8e8';
exports.ccm_chart_N900 = '#ebebeb';
exports.ccm_chart_O100 = '#57330a';
exports.ccm_chart_O200 = '#845117';
exports.ccm_chart_O300 = '#bd7017';
exports.ccm_chart_O400 = '#de8218';
exports.ccm_chart_O50 = '#33210b';
exports.ccm_chart_O500 = '#f2962c';
exports.ccm_chart_O600 = '#f5a54a';
exports.ccm_chart_O700 = '#fabc75';
exports.ccm_chart_O800 = '#ffd8ac';
exports.ccm_chart_O900 = '#ffeace';
exports.ccm_chart_P100 = '#361d61';
exports.ccm_chart_P200 = '#4d2691';
exports.ccm_chart_P300 = '#6533bd';
exports.ccm_chart_P400 = '#7c4ad4';
exports.ccm_chart_P50 = '#281745';
exports.ccm_chart_P500 = '#9762f5';
exports.ccm_chart_P600 = '#a472fc';
exports.ccm_chart_P700 = '#b990ff';
exports.ccm_chart_P800 = '#d3bafd';
exports.ccm_chart_P900 = '#e3d4ff';
exports.ccm_chart_R100 = '#4a1d1b';
exports.ccm_chart_R200 = '#6b2827';
exports.ccm_chart_R300 = '#9e3836';
exports.ccm_chart_R400 = '#cc4743';
exports.ccm_chart_R50 = '#361918';
exports.ccm_chart_R500 = '#f05b56';
exports.ccm_chart_R600 = '#fa7873';
exports.ccm_chart_R700 = '#ff9c99';
exports.ccm_chart_R800 = '#fec6c3';
exports.ccm_chart_R900 = '#ffe0e0';
exports.ccm_chart_T100 = '#153d38';
exports.ccm_chart_T200 = '#23665d';
exports.ccm_chart_T300 = '#269687';
exports.ccm_chart_T400 = '#37b8a6';
exports.ccm_chart_T50 = '#162e2b';
exports.ccm_chart_T500 = '#41ccb9';
exports.ccm_chart_T600 = '#5cdbca';
exports.ccm_chart_T700 = '#7febdc';
exports.ccm_chart_T800 = '#b7f7ef';
exports.ccm_chart_T900 = '#daf9f5';
exports.ccm_chart_V100 = '#4f1c4f';
exports.ccm_chart_V200 = '#7a287a';
exports.ccm_chart_V300 = '#943b94';
exports.ccm_chart_V400 = '#b54ab5';
exports.ccm_chart_V50 = '#3b153b';
exports.ccm_chart_V500 = '#d457d4';
exports.ccm_chart_V600 = '#de73de';
exports.ccm_chart_V700 = '#e898e8';
exports.ccm_chart_V800 = '#f2bff2';
exports.ccm_chart_V900 = '#fbdafb';
exports.ccm_chart_W100 = '#173742';
exports.ccm_chart_W200 = '#225b70';
exports.ccm_chart_W300 = '#2782a3';
exports.ccm_chart_W400 = '#2ea5d1';
exports.ccm_chart_W50 = '#142930';
exports.ccm_chart_W500 = '#42bdeb';
exports.ccm_chart_W600 = '#5fd4fe';
exports.ccm_chart_W700 = '#89dffe';
exports.ccm_chart_W800 = '#c2eeff';
exports.ccm_chart_W900 = '#def6ff';
exports.ccm_chart_Y100 = '#574711';
exports.ccm_chart_Y200 = '#8f7211';
exports.ccm_chart_Y300 = '#c99d08';
exports.ccm_chart_Y400 = '#f0bb0a';
exports.ccm_chart_Y50 = '#382e0c';
exports.ccm_chart_Y500 = '#fac823';
exports.ccm_chart_Y600 = '#fcd456';
exports.ccm_chart_Y700 = '#ffde75';
exports.ccm_chart_Y800 = '#ffe694';
exports.ccm_chart_Y900 = '#fcefc2';
exports.ccm_chart_pareto_label_default = '#ebebeb';
exports.ccm_chart_pareto_line_default = '#faed57';
