// @universe-design/react/es/dialog/dialog/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.u22iagw{display:flex;}
import ukchsq7 from "./../../.stylex/ukchsq7";
//.u6jr5hl{align-items:flex-start;}
import uqt5uie from "./../../.stylex/uqt5uie";
//.u1aehjj5{pointer-events:none;}
import uudhrvd from "./../../.stylex/uudhrvd";
//.utm89h2{max-width:1080px;}
import u16y10k6 from "./../../.stylex/u16y10k6";
//.u7wyk6f{min-width:420px;}
import ua38rkj from "./../../.stylex/ua38rkj";
//.u1rcxdsg{width:calc(100vw - 32px * 2);}
import ulh0qmm from "./../../.stylex/ulh0qmm";
//.u1g96bv0{max-width:840px;}
import u16ivsb9 from "./../../.stylex/u16ivsb9";
//.u177j64f{width:600px;}
import uup8z65 from "./../../.stylex/uup8z65";
//.u1j052i9{min-height:128px;}
import uizwrz6 from "./../../.stylex/uizwrz6";
//.u1qzxk2a{width:420px;}
import ujgul0d from "./../../.stylex/ujgul0d";
//.u113wtx2{top:0px;}
import u1981n0b from "./../../.stylex/u1981n0b";
//.utuwxu6{display:inline-flex;}
import ug0oh3b from "./../../.stylex/ug0oh3b";
//.u1o700av{text-align:left;}
import u135sb2c from "./../../.stylex/u135sb2c";
//.umrv4ls{vertical-align:middle;}
import u1uvgrag from "./../../.stylex/u1uvgrag";
//.u10pi13n{position:relative;}
import u5jxv71 from "./../../.stylex/u5jxv71";
//.u1ewtqcl{box-sizing:border-box;}
import u1p85sbs from "./../../.stylex/u1p85sbs";
//.u4djv7l{top:100px;}
import ul98dlq from "./../../.stylex/ul98dlq";
//.u1hu3pq6{margin-top:0;}
import u57ipeq from "./../../.stylex/u57ipeq";
//.u1ujusj6{margin-right:auto;}
import umejf3v from "./../../.stylex/umejf3v";
//.u19f4twv{margin-bottom:0;}
import u1o4axus from "./../../.stylex/u1o4axus";
//.ucgxt0o{margin-left:auto;}
import uc0fbx2 from "./../../.stylex/uc0fbx2";
//.u6juhto{white-space:normal;}
import u11ie86f from "./../../.stylex/u11ie86f";
//.u17mccla{text-align:center;}
import u1ust6hi from "./../../.stylex/u1ust6hi";
//.uz5stix{white-space:nowrap;}
import uugkpzq from "./../../.stylex/uugkpzq";
//.ud1hv5k::before{display:inline-block;}
import uif4zit from "./../../.stylex/uif4zit";
//.ugovu44::before{width:0px;}
import u1mcpqdr from "./../../.stylex/u1mcpqdr";
//.ur6rhvx::before{height:100%;}
import uq34mcg from "./../../.stylex/uq34mcg";
//.u1o5b37y::before{vertical-align:middle;}
import u5txyzo from "./../../.stylex/u5txyzo";
//.uyl8oag::before{box-sizing:border-box;}
import u4w7uz9 from "./../../.stylex/u4w7uz9";
//.uzhtfnv::before{content:'';}
import u175h1m2 from "./../../.stylex/u175h1m2";
/* eslint-disable max-nested-callbacks */
import stylex from '@universe-design/stylex';
import clsx from 'clsx';
import * as React from 'react';
import { RemoveScrollBar } from 'react-remove-scroll-bar';
import { CSSTransition, Transition } from 'react-transition-group';
import { unstable_useConfigReducedMotion as useConfigReducedMotion } from '../../config-provider/ConfigReducedMotion';
import { createGlobalConfig } from '@universe-design/react-provider';
import { useCssScope } from '../../config-provider/hooks';
import { useIsomorphicLayoutEffect as useLayoutEffect } from '../../hooks';
import Portal from '../../portal';
import { FocusLock } from '../../shared';
import { useAnimatedVisible } from '../../shared/components/transition/useAnimatedVisible';
import { KeyCode, devWarning, getSizeAbbr, getWindow, isNonNullable, parseLength, warning } from '../../util';
import composeRefs from '../../util/composeRefs';
import { forwardHTMLRef } from '../../util/forwardRef';
import { isClient } from '@universe-design/react-utilities';
import '../style';

/**
 * @deprecated
 */

const {
  ConfigProvider,
  useConfig: useDialogLayoutMode,
  setGlobal
} = /* #__PURE__*/createGlobalConfig('dialog-layout-mode', '2.18.0');
export { useDialogLayoutMode };

/**
 * @deprecated
 * 为了更好支持设计规范，我们在 `2.18.0` 修改了对话框的布局方式，由于修改布局方式，对于之前覆盖过对话框样式的用户可能产生 breaking change。默认情况下，我们会使用 `2.18.0` 新布局，如果用户侧存量代码较多，验证成本较高，可以使用该组件修改子树中的对话框布局模式为 `legacy`。
 *
 * 建议渐进式验证并替换为新版 layout，本组件将会在下个大版本中移除
 */
export const DialogLayoutMode = Object.assign(function (_ref) {
  let {
    mode,
    children
  } = _ref;
  return /*#__PURE__*/React.createElement(ConfigProvider, {
    value: mode
  }, children);
}, {
  setGlobal
});
export let DialogCancelType = /*#__PURE__*/function (DialogCancelType) {
  DialogCancelType[DialogCancelType["CancelButton"] = 0] = "CancelButton";
  DialogCancelType[DialogCancelType["CloseIcon"] = 1] = "CloseIcon";
  DialogCancelType[DialogCancelType["Esc"] = 2] = "Esc";
  DialogCancelType[DialogCancelType["Mask"] = 3] = "Mask";
  return DialogCancelType;
}({});
function useDialogCssScope() {
  return useCssScope('dialog');
}
const defaultSafeMarginHmin = 32;
const defaults = {
  getContainer: isClient() ? () => document.body : false,
  safeMargin: {
    top: 32,
    bottom: 32,
    hmax: 180,
    hmin: defaultSafeMarginHmin
  }
};
const defaultSizeCfg = {
  small: {
    minWidth: 420,
    maxWidth: 420
  },
  medium: {
    minWidth: 600,
    maxWidth: 600
  },
  large: {
    minWidth: 420,
    maxWidth: 840
  },
  'extra-large': {
    minWidth: 420,
    maxWidth: 1080
  }
};
const DialogContent = forwardHTMLRef((_ref2, ref) => {
  let {
    centered,
    safeMargin,
    size,
    width: widthProp,
    minWidth: minWidthProp,
    maxWidth: maxWidthProp,
    style: styleProp = {},
    in: inProp,
    ...restProps
  } = _ref2;
  // safeMargin
  const {
    top: safeMarginTop = defaults.safeMargin.top,
    bottom: safeMarginBottom = defaults.safeMargin.bottom,
    hmax: safeMarginHMax = defaults.safeMargin.hmax,
    hmin: safeMarginHMin = defaults.safeMargin.hmin
  } = safeMargin;
  const contentRef = React.useRef(null);
  const mergedContentRef = React.useMemo(() => composeRefs(ref, contentRef), [ref]);

  /**
   * 调整 dialog content 的 top left width height，用于自适应安全边距和最大最小宽度高度。
   * **注意**: 这里 height 相关的实现和技术设计中的描述有一些出入，如果要获取准确的 minHeight 和 maxHeight 需要修改容器的 maxHeight 值再获取 scrollHeight，但是这个过程有两个问题：
   * 1. 会引起 force reflow，会有性能问题 https://gist.github.com/paulirish/5d52fb081b3570c81e3a
   * 2. scrollTop 会失效，需要 reset 后代节点的 scrollTop
   *
   * 为了避免不必要的性能损耗，这里尽可能通过 style 状态和父容器状态进行 nextStyle 的推算，给节点做 nextStyle 赋值，从而走浏览器自己的 batch update 流程。有必要的话可以把这个方法暴露给用户。
   */
  const adjustContentInset = React.useCallback(function () {
    var _contentEl$parentElem, _containerRect$width, _containerRect$height, _ref3, _ref4;
    const contentEl = contentRef.current;
    if (!contentEl) {
      return;
    }
    if (!contentEl.offsetParent) {
      return;
    }
    const computedStyle = getComputedStyle(contentEl);

    // minWidth & maxWidth
    // 优先级 minWidthProp > style.minWidth > computedStyle.minWidth > widthProp > style.width > computedStyle.width > size config
    const minWidth = (() => {
      if (isNonNullable(minWidthProp)) {
        return minWidthProp;
      }
      if (isNonNullable(styleProp.minWidth)) {
        return styleProp.minWidth;
      }
      if (computedStyle.minWidth !== '0px' && !contentEl.style.minWidth) {
        // css minWidth
        return computedStyle.minWidth;
      }
    })();
    const maxWidth = (() => {
      if (isNonNullable(maxWidthProp)) {
        return maxWidthProp;
      }
      if (isNonNullable(styleProp.maxWidth)) {
        return styleProp.maxWidth;
      }
      if (computedStyle.maxWidth !== 'none' && !contentEl.style.maxWidth) {
        return computedStyle.maxWidth;
      }
    })();
    const win = getWindow(contentEl);
    // safari 下 scrollHeight 不准所以用 getBoundingClientRect
    const containerRect = (_contentEl$parentElem = contentEl.parentElement) === null || _contentEl$parentElem === void 0 ? void 0 : _contentEl$parentElem.getBoundingClientRect();
    const containerWidth = (_containerRect$width = containerRect === null || containerRect === void 0 ? void 0 : containerRect.width) !== null && _containerRect$width !== void 0 ? _containerRect$width : win.innerWidth;
    const containerHeight = (_containerRect$height = containerRect === null || containerRect === void 0 ? void 0 : containerRect.height) !== null && _containerRect$height !== void 0 ? _containerRect$height : win.innerHeight;
    const _width = (() => {
      if (isNonNullable(widthProp)) {
        return widthProp;
      }
      if (isNonNullable(styleProp.width)) {
        return styleProp.width;
      }
    })();
    const minWidthNum = parseLength((_ref3 = minWidth !== null && minWidth !== void 0 ? minWidth : _width) !== null && _ref3 !== void 0 ? _ref3 : defaultSizeCfg[size].minWidth, containerWidth);
    const maxWidthNum = parseLength((_ref4 = maxWidth !== null && maxWidth !== void 0 ? maxWidth : _width) !== null && _ref4 !== void 0 ? _ref4 : defaultSizeCfg[size].maxWidth, containerWidth);

    /**
     * 处理宽度
     */
    const {
      width,
      marginH
    } = function () {
      // width 算出来可能是一个非整数，我们不做处理，不然可能会产生 width 抖动导致 reflow
      if (containerWidth < minWidthNum + safeMarginHMin * 2) {
        return {
          width: `${minWidthNum}px`,
          marginH: `${safeMarginHMin}px`
        };
      }
      if (containerWidth < minWidthNum + safeMarginHMax * 2) {
        const margin = (containerWidth - minWidthNum) / 2;
        return {
          width: `${containerWidth - margin * 2}px`,
          marginH: `${margin}px`
        };
      }
      if (containerWidth < maxWidthNum + safeMarginHMax * 2) {
        return {
          width: `${containerWidth - safeMarginHMax * 2}px`,
          marginH: `${safeMarginHMax}px`
        };
      }
      const margin = (containerWidth - maxWidthNum) / 2;
      return {
        width: `${containerWidth - margin * 2}px`,
        marginH: `${margin}px`
      };
    }();

    /**
     * 处理高度
     */
    const {
      height,
      marginTop
    } = function () {
      const _top = (() => {
        if (isNonNullable(styleProp.top)) {
          return styleProp.top;
        }
        if (computedStyle.top !== '0px' && !contentEl.style.top) {
          return computedStyle.top;
        }
        return centered ? 0 : 100;
      })();
      const contentTop = parseLength(_top, containerHeight);
      if (centered) {
        const margin = Math.max(safeMarginTop, safeMarginBottom);
        return {
          height: `${containerHeight - margin * 2}px`,
          marginTop: `${margin}px`
        };
      }
      if (contentTop < safeMarginTop) {
        return {
          height: `${containerHeight - safeMarginTop - safeMarginBottom}px`,
          marginTop: `${safeMarginTop}px`
        };
      }
      return {
        height: `${containerHeight - contentTop - safeMarginBottom}px`,
        marginTop: `${contentTop}px`
      };
    }();

    // 用 marginRight 为了在 wrapper 水平滚动的时候右侧可以撑开，但是很不幸只有最新版的 chrome 是这个行为，其他版本 margin 都是不能占位的。
    // 不用 padding 有两个原因，一是 padding 会导致 mask 无法点击到，需要特殊处理，二是会跟之前的实现有些出入，可能导致用户挂掉
    // TODO marginBottom 在竖直滚动的时候也无法撑开
    contentEl.style.width = width;
    contentEl.style.marginLeft = marginH;
    contentEl.style.marginRight = marginH;
    contentEl.style.height = height;
    contentEl.style.top = marginTop;
  }, [widthProp, styleProp.width, styleProp.minWidth, styleProp.maxWidth, styleProp.top, minWidthProp, size, maxWidthProp, safeMarginHMin, safeMarginHMax, centered, safeMarginTop, safeMarginBottom]);
  useLayoutEffect(() => {
    if (inProp) {
      adjustContentInset();
    }
  }, [adjustContentInset, inProp]);
  React.useEffect(() => {
    const win = getWindow(contentRef.current);
    const cb = () => adjustContentInset();
    win.addEventListener('resize', cb);
    return () => {
      win.removeEventListener('resize', cb);
    };
  });
  return /*#__PURE__*/React.createElement("div", _extends({
    ref: mergedContentRef,
    style: styleProp
  }, restProps));
});
const DialogContentLegacy = forwardHTMLRef((_ref5, ref) => {
  let {
    centered,
    safeMargin,
    ...restProps
  } = _ref5;
  process.env.NODE_ENV !== "production" ? warning(false, devWarning('Dialog',
  // eslint-disable-next-line max-len
  "You are using `Dialog` in legacy mode, legacy mode has been deprecated and will be removed in the next major release. The following properties cannot be used: `minWidth` `maxWidth` `safeMargin.hmin` `safeMargin.hmax` `maxHeight` `bodyHeightResponsiveMode` in legacy mode. Please remove <DialogLayoutMode mode='legacy' /> or wrap your <Dialog /> with <DialogLayoutMode mode='2.18.0' /> for better support for design specs.")) : void 0;
  const {
    top: safeMarginTop = defaults.safeMargin.top,
    bottom: safeMarginBottom = defaults.safeMargin.bottom
  } = safeMargin;
  const contentRef = React.useRef(null);
  React.useImperativeHandle(ref, () => contentRef.current, []);
  useLayoutEffect(() => {
    // 矫正 content top 不超过安全边距
    if (contentRef.current) {
      let contentTop = parseFloat(getComputedStyle(contentRef.current).top || '0');
      if (!centered && contentTop < safeMarginTop) {
        contentTop = safeMarginTop;
        contentRef.current.style.top = `${safeMarginTop}px`;
      } else if (centered && contentTop > 0) {
        // top 属性失效
        contentTop = 0;
        contentRef.current.style.top = '0px';
      }
      contentRef.current.style.maxHeight = `calc(100vh - ${centered ? safeMarginTop + safeMarginBottom : safeMarginBottom + contentTop}px)`;
    }
  });
  return /*#__PURE__*/React.createElement("div", _extends({
    ref: contentRef
  }, restProps));
});

// eslint-disable-next-line max-lines-per-function
export const Dialog = /*#__PURE__*/React.forwardRef((props, ref) => {
  const {
    visible = false,
    overlay = true,
    overlayClosable,
    keyboardClosable = true,
    contentStyle,
    overlayStyle,
    overlayClassName,
    contentClassName,
    wrapperClassName,
    enableAnimation: enableAnimationProp,
    overlayTransitionClassName,
    contentTransitionClassName,
    destroyOnClose,
    children,
    width,
    zIndex,
    centered = true,
    getContainer = defaults.getContainer,
    onClose,
    onAfterClose,
    onAfterOpen,
    size = 'medium',
    safeMargin = defaults.safeMargin,
    className,
    initialFocusRef,
    finalFocusRef,
    enableFocusLock = false,
    enableRestoreFocus = false,
    forceRender,
    minWidth,
    maxWidth,
    ...restProps
  } = props;
  const dialogCssScope = useDialogCssScope();
  const layoutMode = useDialogLayoutMode();
  const enableAnimation = !useConfigReducedMotion(isNonNullable(enableAnimationProp) ? !enableAnimationProp : undefined);
  const wrapperRef = React.useRef(null);
  const contentRef = React.useRef(null);
  const contentClickRef = React.useRef();
  const contentTimeoutRef = React.useRef();

  // 添加 esc 监听
  const onKeyDown = React.useCallback(e => {
    if (keyboardClosable) {
      const {
        keyCode
      } = e;
      /* istanbul ignore else  */
      if (keyCode === KeyCode.ESC) {
        onClose && onClose(e, {
          cancelType: DialogCancelType.Esc
        });
      }
    }
  }, [keyboardClosable, onClose]);
  const handleOverlayClick = e => {
    if (!overlayClosable) {
      return;
    }
    if (contentClickRef.current) {
      contentClickRef.current = false;
    } else if (wrapperRef.current === e.target) {
      onClose && onClose(e, {
        cancelType: DialogCancelType.Mask
      });
    }
  };
  const onContentMouseDown = () => {
    window.clearTimeout(contentTimeoutRef.current);
    contentClickRef.current = true;
  };
  const onContentMouseUp = () => {
    contentTimeoutRef.current = window.setTimeout(() => {
      contentClickRef.current = false;
    });
  };
  const {
    transitionGroupProps,
    transitionGroupPropsWithoutListener,
    styleDisplay,
    noopRef
  } = useAnimatedVisible({
    destroyOnClose,
    forceRender,
    visible,
    onAfterClose,
    onAfterOpen
  });
  const timeout = React.useMemo(() => ({
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    enter: enableAnimation ? 400 : 0,
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    exit: enableAnimation ? 300 : 0
  }), [enableAnimation]);
  const maskRef = React.useRef(null);
  const mergedInitialFocusRef = initialFocusRef === null ? null : initialFocusRef !== null && initialFocusRef !== void 0 ? initialFocusRef : wrapperRef;
  const renderOverlay = () => {
    if (!overlay) {
      return null;
    }
    const cls = clsx(dialogCssScope('__mask'), overlayClassName);
    let animationCls = '';
    if (enableAnimation) {
      animationCls = overlayTransitionClassName || dialogCssScope('__mask-animation');
    }
    const mergeStyle = {
      ...overlayStyle,
      ...styleDisplay
    };
    if (zIndex || zIndex === 0) {
      mergeStyle.zIndex = zIndex;
    }
    return /*#__PURE__*/React.createElement(CSSTransition, _extends({
      timeout: timeout,
      appear: true,
      classNames: animationCls,
      nodeRef: maskRef
    }, transitionGroupPropsWithoutListener), /*#__PURE__*/React.createElement("div", {
      ref: maskRef,
      style: mergeStyle,
      className: cls
    }));
  };
  const usedSize = getSizeAbbr(size);
  const renderContent = () => {
    const wrapCls = clsx(dialogCssScope('__wrap'), wrapperClassName, {
      [dialogCssScope('--centered')]: centered
    }, stylex(layoutMode === 'legacy' ? [centered && {
      "props": {
        "fsow6f": u1ust6hi,
        "Huce71": uugkpzq,
        "Br0sdwz": uif4zit,
        "Ccq8qp": u1mcpqdr,
        "Baz25je": uq34mcg,
        "Byamso": u5txyzo,
        "Bcvre1j": u4w7uz9,
        "Ftih45": u175h1m2
      }
    }] : [centered && {
      "props": {
        "Huce71": uugkpzq
      }
    }]));
    const cls = clsx(dialogCssScope('__content'), contentClassName, {
      [dialogCssScope(`__content-${usedSize}`)]: usedSize
    }, stylex(layoutMode === 'legacy' ? [{
      "props": {
        "qhf8xq": u5jxv71,
        "B7ck84d": u1p85sbs,
        "Bhzewxz": ul98dlq,
        "B6of3ja": u57ipeq,
        "t21cq0": umejf3v,
        "jrapky": u1o4axus,
        "Frg6f3": uc0fbx2,
        "Huce71": u11ie86f
      }
    }, centered && {
      "props": {
        "Bhzewxz": u1981n0b,
        "mc9l5x": ug0oh3b,
        "fsow6f": u135sb2c,
        "ha4doy": u1uvgrag
      }
    }, usedSize && {
      sm: {
        "props": {
          "a9b677": ujgul0d
        }
      },
      md: {
        "props": {
          "a9b677": uup8z65,
          "sshi5w": uizwrz6
        }
      },
      lg: {
        "props": {
          "Bf4jedk": ua38rkj,
          "a9b677": ulh0qmm,
          "B2u0y6b": u16ivsb9
        }
      },
      xl: {
        "props": {
          "Bf4jedk": ua38rkj,
          "a9b677": ulh0qmm,
          "B2u0y6b": u16y10k6
        }
      }
    }[usedSize]] : [{
      "props": {
        "qhf8xq": u5jxv71,
        "Huce71": u11ie86f,
        "mc9l5x": ukchsq7,
        "Bt984gj": uqt5uie,
        "Bkecrkj": uudhrvd
      }
    }]));
    const wrapStyle = {
      ...styleDisplay
    };
    if (zIndex || zIndex === 0) {
      wrapStyle.zIndex = zIndex;
    }
    let animationCls = '';
    if (enableAnimation) {
      animationCls = contentTransitionClassName || dialogCssScope('__content-animation');
    }
    return /*#__PURE__*/React.createElement(CSSTransition, _extends({
      timeout: timeout,
      appear: true,
      classNames: animationCls,
      nodeRef: contentRef // 这里改为 contentRef 因为在用 ScrollArea 的时候和之前用 div 的时候行为不同，关闭 dialog 会有偶现的闪动
    }, transitionGroupPropsWithoutListener), /*#__PURE__*/React.createElement(FocusLock, {
      isDisabled: !visible,
      initialFocusRef: mergedInitialFocusRef,
      restoreFocus: enableRestoreFocus,
      finalFocusRef: finalFocusRef,
      enableFocusLock: enableFocusLock
    }, /*#__PURE__*/React.createElement("div", {
      className: wrapCls,
      style: wrapStyle,
      tabIndex: -1,
      onClick: handleOverlayClick,
      onKeyDown: onKeyDown,
      ref: wrapperRef
    }, layoutMode === 'legacy' ? /*#__PURE__*/React.createElement(DialogContentLegacy, {
      style: {
        ...contentStyle,
        width
      },
      className: cls,
      onMouseDown: onContentMouseDown,
      onMouseUp: onContentMouseUp,
      safeMargin: safeMargin,
      centered: centered,
      role: "dialog",
      "aria-modal": "true",
      ref: contentRef
    }, children) : /*#__PURE__*/React.createElement(DialogContent, {
      style: contentStyle,
      className: cls,
      onMouseDown: onContentMouseDown,
      onMouseUp: onContentMouseUp,
      safeMargin: safeMargin,
      centered: centered,
      role: "dialog",
      "aria-modal": "true",
      ref: contentRef,
      size: size,
      width: width,
      minWidth: minWidth,
      maxWidth: maxWidth,
      in: transitionGroupPropsWithoutListener.in
    }, children))));
  };
  return /*#__PURE__*/React.createElement(Transition, _extends({
    timeout: timeout,
    appear: true,
    nodeRef: noopRef
  }, transitionGroupProps), (() => {
    const dialogRootNode = /*#__PURE__*/React.createElement("div", _extends({
      ref: ref,
      className: clsx(dialogCssScope('__root'), className)
    }, restProps), visible && /*#__PURE__*/React.createElement(RemoveScrollBar, {
      noRelative: true,
      gapMode: "margin"
    }), renderOverlay(), renderContent());
    if (getContainer === false) {
      return dialogRootNode;
    }
    return /*#__PURE__*/React.createElement(Portal, {
      container: getContainer
    }, dialogRootNode);
  })());
});
