// @universe-design/react/lib/upload/hooks/useUpload.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useUpload = useUpload;
var React = _interopRequireWildcard(require("react"));
var _uuid = require("uuid");
var _configProvider = require("../../config-provider");
var _useMergedState = _interopRequireDefault(require("../../hooks/useMergedState"));
var _utils = require("../utils");
var _useEventCallback = require("../../hooks/useEventCallback");
var _hooks = require("../../hooks");
var _useUploadTrigger = require("./useUploadTrigger");
function useUpload(props, uploadRef) {
  const {
    fileList,
    defaultFileList,
    prefixCls,
    onChange: onChangeProp,
    disabled = false,
    maxCount,
    onExceed
  } = props;

  /**
   * ====================
   *   ClassName
   * ====================
   */
  const rootPrefixCls = (0, _configProvider.useCssScope)('upload', prefixCls)();

  /**
   * ====================
   *   File List
   * ====================
   */
  const defaultFileListFormatted = (defaultFileList || []).map(file => ({
    ...file,
    uid: file.uid ? file.uid : (0, _uuid.v4)(),
    status: file.status ? file.status : 'success'
  }));
  const [mergedFileList, setMergedFileList] = (0, _useMergedState.default)(defaultFileListFormatted, {
    value: fileList,
    postState: list => list !== null && list !== void 0 ? list : []
  });
  const fileListRef = (0, _hooks.useLatest)(mergedFileList);

  // 为文件生成默认的uid/status
  React.useMemo(() => {
    fileList === null || fileList === void 0 ? void 0 : fileList.forEach(file => {
      if (Object.isFrozen(file)) {
        return;
      }
      if (!file.status) {
        file.status = 'success';
      }
      if (!file.uid) {
        file.uid = (0, _uuid.v4)();
      }
    });
  }, [fileList]);
  const onChange = (0, _useEventCallback.useEventCallback)((file, changedFileList, event) => {
    let cloneList = [...changedFileList];
    if (maxCount === 1) {
      cloneList = cloneList.slice(-1);
    } else if (maxCount) {
      cloneList = cloneList.slice(0, maxCount);
    }
    setMergedFileList(cloneList);
    fileListRef.current = cloneList;
    const changeInfo = {
      file: file,
      fileList: cloneList
    };
    if (event) {
      changeInfo.event = event;
    }
    onChangeProp === null || onChangeProp === void 0 ? void 0 : onChangeProp(changeInfo);
  });

  /**
   * ====================
   *   Upload Trigger
   * ====================
   */
  const {
    uploadFiles,
    abort,
    retry,
    onBatchStart,
    onError,
    onProgress,
    onSuccess
  } = (0, _useUploadTrigger.useUploadTrigger)(props, {
    fileListRef,
    onChange
  });

  /**
   * ====================
   *   Upload File
   * ====================
   */
  const onRemove = React.useCallback(file => {
    let currentFile;
    const removedFileList = (0, _utils.removeFileItem)(file, mergedFileList);
    if (removedFileList) {
      currentFile = {
        ...file,
        status: 'removed'
      };
      mergedFileList.forEach(item => {
        const matchKey = currentFile.uid !== undefined ? 'uid' : 'name';
        if (item[matchKey] === currentFile[matchKey] && !Object.isFrozen(item)) {
          item.status = 'removed';
        }
      });
      abort(currentFile);
      onChange(currentFile, removedFileList);
    }
  }, [abort, mergedFileList, onChange]);
  const onRetry = React.useCallback(file => {
    if (file.originFileObj) {
      // 文件已损坏
      retry(file.originFileObj);
    }
  }, [retry]);

  /**
   * ====================
   *   Upload Ref
   * ====================
   * 用于测试
   */
  React.useImperativeHandle(uploadRef, () => ({
    onBatchStart,
    onSuccess,
    onProgress,
    onError,
    fileList: mergedFileList,
    upload: {
      uploadFiles,
      abort,
      retry
    }
  }));
  return {
    onChange,
    fileList: mergedFileList,
    disabled,
    rootPrefixCls,
    maxCount,
    onExceed,
    onRemove,
    onRetry,
    fileListRef,
    uploadFiles
  };
}
