// @byted-service/metrics/lib/lfu/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LFU = void 0;
const list_1 = require("./list");
const item_1 = require("./item");
const node_1 = require("./node");
const node_value_1 = require("./node_value");
class LFU {
    constructor(capacity) {
        this.kv = new Map();
        this.freqList = new list_1.List();
        this.capacity = capacity;
    }
    set(k, value) {
        if (this.kv.size >= this.capacity) {
            this.evict();
        }
        if (this.kv.has(k)) {
            const item = this.kv.get(k);
            item.v = value;
            this.increase(item);
            return;
        }
        const item = new item_1.Item(k, value);
        const header = this.freqList.header;
        if (!header || header.value.freq !== 1) {
            const nodeValue = new node_value_1.NodeValue(1);
            const node = new node_1.Node(nodeValue);
            item.freqNode = node;
            this.freqList.insertHeader(node);
            nodeValue.items.set(item, true);
        }
        else {
            item.freqNode = header;
            header.value.items.set(item, true);
        }
        this.kv.set(k, item);
    }
    get(key) {
        const val = this.kv.get(key);
        if (val) {
            this.increase(val);
        }
        return val === null || val === void 0 ? void 0 : val.v;
    }
    evict() {
        if (this.freqList.length === 0) {
            return false;
        }
        const header = this.freqList.header;
        const keys = header.value.items.keys();
        const key = keys.next().value;
        header.value.items.delete(key);
        this.kv.delete(key.k);
        if (header.value.items.size === 0) {
            this.freqList.remove(header);
        }
        return true;
    }
    dump() {
        const arr = [];
        this.kv.forEach((val, key) => {
            arr.push({ key, val: val.v, freq: val.freqNode.value.freq });
        });
        return arr;
    }
    increase(item) {
        const freqNode = item.freqNode;
        const next = freqNode.next;
        if (!next || next.value.freq !== freqNode.value.freq + 1) {
            const nodeValue = new node_value_1.NodeValue(freqNode.value.freq + 1);
            nodeValue.items.set(item, true);
            const node = new node_1.Node(nodeValue);
            this.freqList.insertAfter(node, freqNode);
            item.freqNode = node;
        }
        else {
            next.value.items.set(item, true);
            item.freqNode = next;
        }
        freqNode.value.items.delete(item);
        if (freqNode.value.items.size === 0) {
            this.freqList.remove(freqNode);
        }
    }
}
exports.LFU = LFU;
