// @byted-service/metrics/lib/lfu/list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.List = void 0;
// double linked list
class List {
    constructor() {
        this.header = null;
        this.length = 0;
    }
    insertHeader(node) {
        if (!this.header) {
            this.header = node;
        }
        else {
            node.next = this.header;
            this.header.prev = node;
            this.header = node;
        }
        this.length++;
    }
    /* istanbul ignore next */
    insertBefore(node, cur) {
        const prev = cur.prev;
        cur.prev = node;
        node.next = cur;
        node.prev = prev;
        if (prev) {
            prev.next = node;
        }
        this.length++;
    }
    insertAfter(node, cur) {
        const next = cur.next;
        cur.next = node;
        node.prev = cur;
        node.next = next;
        if (next) {
            next.prev = node;
        }
        this.length++;
    }
    remove(node) {
        if (node === this.header) {
            this.header = this.header.next;
            if (this.header) {
                this.header.prev = null;
            }
        }
        else {
            const prev = node.prev;
            const next = node.next;
            prev.next = next;
            next.prev = prev;
        }
        this.length--;
    }
}
exports.List = List;
