// @byted-service/netpool/lib/dynamic/pool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DynamicPool = void 0;
const helper_1 = require("../helper");
const consts_1 = require("../consts");
const usingCountSym = Symbol.for('usingCount');
const lastUseTimeSym = Symbol.for('lastUseTime');
const poolSym = Symbol.for('pool');
const idCreator = (0, helper_1.genIdCreator)('dynamicpool');
/**
 * Global clock
 * 全局时钟
 */
let now = Date.now();
setInterval(() => (now = Date.now()), 1000).unref();
class DynamicPool {
    constructor(options) {
        this.id = options.id || idCreator();
        this.poolCache = options.poolCache;
        this.logger = options.logger || consts_1.defaultLogger;
        this.checkInterval =
            options.checkInterval || consts_1.defaultDynamicPoolOptions.checkInterval;
        this.maxFreeTime =
            options.maxFreeTime || consts_1.defaultDynamicPoolOptions.maxFreeTime;
        this.checkPoolStatus();
    }
    async acquire(options, ...args) {
        const pool = this.poolCache.get(options, ...args);
        const resource = await pool.acquire(...args);
        resource[poolSym] = pool;
        pool[usingCountSym] = (pool[usingCountSym] || 0) + 1;
        pool[lastUseTimeSym] = now;
        return resource;
    }
    release(resource) {
        const pool = resource[poolSym];
        if (!pool) {
            throw new Error('pool symbol not exists');
        }
        pool[usingCountSym]--;
        pool[lastUseTimeSym] = now;
        return pool.release(resource);
    }
    async destroy() {
        clearTimeout(this.checkIntervalId);
        for (const pool of this.poolCache.values()) {
            this.poolCache.delete(pool);
            await pool.destroy();
        }
    }
    async checkPoolStatus() {
        for (const pool of this.poolCache.values()) {
            if (pool[usingCountSym] <= 0 &&
                pool[lastUseTimeSym] < Date.now() - this.maxFreeTime) {
                try {
                    this.poolCache.delete(pool);
                    await pool.destroy();
                }
                catch (err) {
                    this.logger.warn('[netpool] [%s] Failed to destroy pool "%s". %o', this.id, pool.id || '-', err);
                }
            }
        }
        this.checkIntervalId = setTimeout(this.checkPoolStatus.bind(this), this.checkInterval).unref();
    }
}
exports.DynamicPool = DynamicPool;
