// @universe-design/react/lib/date-picker/weekNumberWrapper/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _tooltip = require("../../tooltip");
var _hooks = require("../../config-provider/hooks");
require("./index.less");
const WeekNumberWrapper = /*#__PURE__*/_react.default.forwardRef((_ref, ref) => {
  let {
    hideWeekNumbers,
    rowNum,
    colNum,
    baseDate,
    children,
    ...restProps
  } = _ref;
  const cssScope = (0, _hooks.useCssScope)('picker__cell-week-number');
  const wrapperCssScope = (0, _hooks.useCssScope)('picker__wrapper-week-number');
  const headerCssScope = (0, _hooks.useCssScope)('picker__header-week-number');
  const {
    components: {
      datePicker: locale
    }
  } = (0, _hooks.useLocale)();
  if (hideWeekNumbers) {
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, children);
  }
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
    ref: ref,
    className: (0, _clsx.default)(wrapperCssScope())
  }, restProps), /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(headerCssScope())
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(headerCssScope('-name'))
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(headerCssScope('-content'))
  }, (() => [...new Array(rowNum)].map((_, rowIndex) => {
    const weekEndDate = baseDate.clone().add((rowIndex + 1) * colNum - 1, 'day');
    const weekNumber = weekEndDate.weeks();
    const tooltipTitle = locale.weekNumber.f({
      num: String(weekNumber)
    });
    return /*#__PURE__*/_react.default.createElement("div", {
      className: (0, _clsx.default)(cssScope()),
      key: rowIndex
    }, /*#__PURE__*/_react.default.createElement(_tooltip.Tooltip, {
      title: tooltipTitle
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: (0, _clsx.default)(cssScope('__inner'))
    }, weekNumber)));
  }))())), /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(headerCssScope('__space'))
  }), children);
});
var _default = exports.default = WeekNumberWrapper;
