// @universe-design/react/lib/pro/timezone-picker/const.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "TIMEZONE_LINK_MAP", {
  enumerable: true,
  get: function () {
    return _timezoneLink.TIMEZONE_LINK_MAP;
  }
});
exports.getTimezoneMetaMap = exports.getSortedTimezoneMetaList = exports.extendRawMeta = exports.TIMEZONE_META_LIST = void 0;
var _momentTimezone = _interopRequireDefault(require("@lark/moment-timezone"));
var _keyBy = _interopRequireDefault(require("lodash/keyBy"));
var _once = _interopRequireDefault(require("lodash/once"));
var _timezoneMeta = require("./timezoneMeta");
var _timezoneLink = require("./timezoneLink");
const getTimezoneLocaleKey = timezone => {
  const lastSlashPos = timezone.lastIndexOf('/');
  return timezone.slice(lastSlashPos + 1);
};
const extendRawMeta = rawMeta => ({
  ...rawMeta,
  l: `Tz_${getTimezoneLocaleKey(rawMeta.v)}`,
  s: Object.entries(rawMeta.c).map(_ref => {
    let [key, value] = _ref;
    return Object.assign([key, ...value], {
      c: `Ct_${key}`
    });
  })
});
exports.extendRawMeta = extendRawMeta;
const TIMEZONE_META_LIST = exports.TIMEZONE_META_LIST = _timezoneMeta.TIMEZONE_RAW_META_LIST.map(extendRawMeta);
const timezoneSorter = (a, b) => {
  const offsetA = (0, _momentTimezone.default)().tz(a.v).utcOffset();
  const offsetB = (0, _momentTimezone.default)().tz(b.v).utcOffset();
  return offsetA - offsetB;
};

/** 数组按照 GMT 偏移量从小到大排序 */
const getSortedTimezoneMetaList = exports.getSortedTimezoneMetaList = (0, _once.default)(() => [...TIMEZONE_META_LIST].sort(timezoneSorter));

/** 获取 timezone meta map */
const getTimezoneMetaMap = exports.getTimezoneMetaMap = (0, _once.default)(() => (0, _keyBy.default)(TIMEZONE_META_LIST, item => item.v));
