// @universe-design/react/lib/pro/timezone-picker/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "TimezoneLabel", {
  enumerable: true,
  get: function () {
    return _TimezoneLabel.TimezoneLabel;
  }
});
exports.TimezonePicker = void 0;
Object.defineProperty(exports, "chinaMainlandMetaTransformer", {
  enumerable: true,
  get: function () {
    return _utils.chinaMainlandMetaTransformer;
  }
});
Object.defineProperty(exports, "getTimezoneFormatData", {
  enumerable: true,
  get: function () {
    return _utils.getTimezoneFormatData;
  }
});
Object.defineProperty(exports, "getTimezoneNames", {
  enumerable: true,
  get: function () {
    return _utils.getTimezoneNames;
  }
});
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _momentTimezone = _interopRequireDefault(require("@lark/moment-timezone"));
var _isEmpty = _interopRequireDefault(require("lodash/isEmpty"));
var _hooks = require("../../config-provider/hooks");
var _select = require("../../select");
var _text = require("../../text");
var _reactUtilities = require("@universe-design/react-utilities");
var _highlighter = require("../../shared/components/highlighter");
var _common = require("../../select/common");
var _utils = require("./utils");
require("./style");
var _TimezoneLabel = require("./TimezoneLabel");
const TimezonePicker = exports.TimezonePicker = /*#__PURE__*/React.memo(props => {
  const defaultMetaTransformer = (0, _utils.useMetaTransformer)();
  const {
    value: valueProp,
    defaultValue: defaultValueProp,
    date,
    metaTransformer = defaultMetaTransformer,
    ...restProps
  } = props;
  const dateString = date === null || date === void 0 ? void 0 : date.toString();

  /**
   * Context
   */
  const cssScope = (0, _common.useSelectListItemCssScope)();
  const timezonePickerCssScope = (0, _hooks.useCssScope)('timezone-picker');

  /**
   * Locale
   */
  const locale = (0, _hooks.useLocale)();
  const timezoneLocale = (0, _utils.useTimezoneLocale)();

  /**
   * Value
   */
  const value = React.useMemo(() => {
    var _getTimezoneMeta$v, _getTimezoneMeta;
    return (_getTimezoneMeta$v = (_getTimezoneMeta = (0, _utils.getTimezoneMeta)(valueProp)) === null || _getTimezoneMeta === void 0 ? void 0 : _getTimezoneMeta.v) !== null && _getTimezoneMeta$v !== void 0 ? _getTimezoneMeta$v : valueProp;
  }, [valueProp]);

  // 线上逻辑，兼容时区不完整
  const defaultValue = (0, _reactUtilities.useConstant)(() => {
    var _getTimezoneMeta$v2, _getTimezoneMeta2;
    return (_getTimezoneMeta$v2 = (_getTimezoneMeta2 = (0, _utils.getTimezoneMeta)(defaultValueProp)) === null || _getTimezoneMeta2 === void 0 ? void 0 : _getTimezoneMeta2.v) !== null && _getTimezoneMeta$v2 !== void 0 ? _getTimezoneMeta$v2 : defaultValueProp;
  });

  /**
   * Search
   */
  const [searchValue, setSearchValue] = React.useState('');
  const handleSearch = React.useCallback(inputValue => {
    setSearchValue(inputValue);
  }, []);

  /**
   * Options
   */

  // 原始 meta list 的 text 和 search 字段只是占位符，需要做一下「fill in」操作
  const filledTimezoneMetaList = React.useMemo(() => {
    const _date = dateString ? (0, _momentTimezone.default)(dateString) : undefined;
    return timezoneLocale ? (0, _utils.fillInTimezoneMetaList)(timezoneLocale, _date, metaTransformer) : [];
  }, [timezoneLocale, dateString, metaTransformer]);
  const options = React.useMemo(() => {
    // TODO 之后把 matcher 能力和 scorer 能力也放进去，label 直接基于 matches 数组去做渲染就行了
    const filteredTimezoneMetaList = (0, _utils.searchTimezoneMetaList)(filledTimezoneMetaList, searchValue);
    const searchWords = (0, _utils.splitKeywords)(searchValue);
    const highlightClassName = cssScope('__content-highlight');
    const _options = filteredTimezoneMetaList.map(timezoneMeta => {
      var _timezoneMeta$cities;
      return {
        value: timezoneMeta.v,
        title: timezoneMeta.l,
        label: /*#__PURE__*/React.createElement("div", {
          className: timezonePickerCssScope('__label')
        }, /*#__PURE__*/React.createElement(_highlighter.Highlighter, {
          searchWords: searchWords,
          textToHighlight: timezoneMeta.l,
          markRenderer: text => /*#__PURE__*/React.createElement("span", {
            className: highlightClassName
          }, text)
        }), (0, _isEmpty.default)(timezoneMeta.cities) ? null : /*#__PURE__*/React.createElement(_text.Text, {
          as: "div",
          className: timezonePickerCssScope('__subtitle'),
          lineClamp: 2
        }, /*#__PURE__*/React.createElement("span", null, timezoneLocale === null || timezoneLocale === void 0 ? void 0 : timezoneLocale.Ct_Includes), (0, _utils.arrayJoin)((_timezoneMeta$cities = timezoneMeta.cities) === null || _timezoneMeta$cities === void 0 ? void 0 : _timezoneMeta$cities.map((city, index) => /*#__PURE__*/React.createElement("span", {
          key: `item_${index}`,
          className: highlightClassName
        }, city)), i => /*#__PURE__*/React.createElement("span", {
          key: `joiner_${i}`
        }, "\u3001"))))
      };
    });
    return _options;
  }, [timezoneLocale, filledTimezoneMetaList, searchValue, cssScope, timezonePickerCssScope]);
  return /*#__PURE__*/React.createElement(_select.Select
  // 由于这里是手动控制 options，所以用 noOptionsContent 而不是 notFoundContent
  , (0, _extends2.default)({
    noOptionsContent: locale.global.notFoundContent
  }, restProps, 'value' in props ? {
    value
  } : {}, {
    defaultValue,
    filterOption: false,
    searchValue,
    onSearch: handleSearch,
    options
  }));
});
