// @universe-design/react/lib/pro/timezone-picker/raw-data.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.timezoneCityRelations = void 0;
/** 内置中文和拼音 */

/** 可被搜索的数据 */

const timezoneCityRelations = exports.timezoneCityRelations = {
  'Pacific/Niue': {
    Niue: ['纽埃', 'niu ai']
  },
  'Pacific/Pago_Pago': {
    Samoa: ['萨摩亚', 'sa mo ya']
  },
  'Pacific/Rarotonga': {
    CookIslands: ['库克群岛', 'ku ke qun dao']
  },
  'Pacific/Honolulu': {
    Hawaii: ['夏威夷', 'xia wei yi'],
    Honolulu: ['檀香山', 'tan xiang shan']
  },
  'Pacific/Tahiti': {
    Tahiti: ['塔希提岛', 'ta xi ti dao']
  },
  'Pacific/Marquesas': {
    Marquesas: ['马克萨斯群岛', 'ma ke sa si qun dao']
  },
  'America/Adak': {
    AleutianIslands: ['阿留申群岛', 'a liu shen qun dao']
  },
  'Pacific/Gambier': {
    Gambier: ['甘比尔', 'gan bi er']
  },
  'America/Anchorage': {
    Anchorage: ['安克雷奇', 'an ke lei qi']
  },
  'America/Juneau': {
    Juneau: ['朱诺', 'zhu nuo']
  },
  'America/Metlakatla': {
    Metlakatla: ['梅特拉卡特拉', 'mei te la ka te la']
  },
  'America/Nome': {
    Nome: ['诺姆', 'nuo mu']
  },
  'America/Sitka': {
    Sitka: ['锡特卡', 'xi te ka']
  },
  'America/Yakutat': {
    Yakutat: ['亚库塔特', 'ya ku ta te']
  },
  'Pacific/Pitcairn': {
    Pitcairn: ['皮特凯恩', 'pi te kai en']
  },
  'America/Los_Angeles': {
    LosAngeles: ['洛杉矶', 'luo shan ji'],
    SanFrancisco: ['旧金山', 'jiu jin shan'],
    SanDiego: ['圣迭戈', 'sheng die ge'],
    Seattle: ['西雅图', 'xi ya tu'],
    LasVegas: ['拉斯维加斯', 'la si wei jia si'],
    SanJose: ['圣何塞', 'sheng he sai'],
    Portland: ['波特兰', 'bo te lan'],
    Riverside: ['河滨', 'he bin'],
    Sacramento: ['萨克拉门托', 'sa ke la men tuo'],
    MountainView: ['山景城', 'shan jing cheng'],
    Irvine: ['尔湾', 'er wan']
  },
  'America/Tijuana': {
    Tijuana: ['蒂华纳', 'di hua na']
  },
  'America/Vancouver': {
    Vancouver: ['温哥华', 'wen ge hua']
  },
  'America/Dawson': {
    Dawson: ['道森', 'dao sen']
  },
  'America/Whitehorse': {
    Whitehorse: ['怀特霍斯', 'huai te huo si']
  },
  'America/Hermosillo': {
    Hermosillo: ['埃莫西约', 'ai mo xi yue']
  },
  'America/Chihuahua': {
    Chihuahua: ['奇瓦瓦', 'qi wa wa']
  },
  'America/Mazatlan': {
    Mazatlan: ['马萨特兰', 'ma sa te lan']
  },
  'America/Creston': {
    Creston: ['克雷斯顿', 'ke lei si dun']
  },
  'America/Dawson_Creek': {
    DawsonCreek: ['道森克里克', 'dao sen ke li ke']
  },
  'America/Fort_Nelson': {
    FortNelson: ['纳尔逊堡', 'na er xun bao']
  },
  'America/Phoenix': {
    Phoenix: ['菲尼克斯', 'fei ni ke si']
  },
  'America/Boise': {
    Boise: ['博伊西', 'bo yi xi']
  },
  'America/Cambridge_Bay': {
    CambridgeBay: ['剑桥湾', 'jian qiao wan']
  },
  'America/Denver': {
    Denver: ['丹佛', 'dan fo'],
    SaltLakeCity: ['盐湖城', 'yan hu cheng']
  },
  'America/Edmonton': {
    Edmonton: ['埃德蒙顿', 'ai de meng dun'],
    Calgary: ['卡尔加里', 'ka er jia li']
  },
  'America/Inuvik': {
    Inuvik: ['伊努维克', 'yi nu wei ke']
  },
  'America/Ojinaga': {
    Ojinaga: ['奥希纳加', 'ao xi na jia']
  },
  'America/Yellowknife': {
    Yellowknife: ['耶洛奈夫', 'ye luo nai fu']
  },
  'America/Belize': {
    Belize: ['伯利兹', 'bo li zi']
  },
  'America/Costa_Rica': {
    CostaRica: ['哥斯达黎加', 'ge si da li jia']
  },
  'America/El_Salvador': {
    ElSalvador: ['萨尔瓦多', 'sa er wa duo'],
    SanSalvador: ['圣萨尔瓦多', 'sheng sa er wa duo']
  },
  'America/Guatemala': {
    Guatemala: ['危地马拉', 'wei di ma la'],
    GuatemalaCity: ['危地马拉城', 'wei de ma la cheng']
  },
  'America/Managua': {
    Managua: ['马那瓜', 'ma na gua']
  },
  'America/Regina': {
    Regina: ['里贾纳', 'li jia na']
  },
  'America/Swift_Current': {
    SwiftCurrent: ['斯威夫特卡伦特', 'si wei fu te ka lun te']
  },
  'America/Tegucigalpa': {
    Tegucigalpa: ['特古西加尔巴', 'te gu xi jia er ba']
  },
  'America/Bahia_Banderas': {
    BahiadeBanderas: ['巴伊亚班德拉斯', 'ba yi ya ban de la si']
  },
  'America/Merida': {
    Merida: ['梅里达', 'mei li da']
  },
  'America/Mexico_City': {
    MexicoCity: ['墨西哥城', 'mo xi ge cheng'],
    Guadalajara: ['瓜达拉哈拉', 'gua da la ha la'],
    Puebla: ['普埃布拉', 'pu ai bu la']
  },
  'America/Monterrey': {
    Monterrey: ['蒙特雷', 'meng te lei']
  },
  'Pacific/Galapagos': {
    Galapagos: ['加拉帕戈斯', 'jia la pa ge si']
  },
  'America/North_Dakota/Beulah': {
    NorthDakotaBeulah: ['北达科他州比尤拉', 'bei da ke ta zhou bi you la']
  },
  'America/North_Dakota/Center': {
    NorthDakotaCenterCity: ['北达科他州申特城', 'bei da ke ta zhou shen te cheng']
  },
  'America/Chicago': {
    Chicago: ['芝加哥', 'zhi jia ge'],
    Houston: ['休斯敦', 'xiu si dun'],
    Dallas: ['达拉斯', 'da la si'],
    KansasCity: ['堪萨斯城', 'kan sa si cheng'],
    Minneapolis: ['明尼阿波利斯', 'ming ni a bo li si'],
    Austin: ['奥斯汀', 'ao si ting'],
    NewOrleans: ['新奥尔良', 'xin ao er liang'],
    Nashville: ['纳什维尔', 'na shi wei er'],
    PortLouis: ['路易港', 'lu yi gang'],
    Birmingham: ['伯明翰', 'bo ming han'],
    SanAntonio: ['圣安东尼奥', 'sheng an dong ni ao'],
    Milwaukee: ['密尔沃基', 'mi er wo ji'],
    OklahomaCity: ['俄克拉荷马城', 'e ke la he ma cheng'],
    Memphis: ['孟菲斯', 'meng fei si']
  },
  'America/Indiana/Knox': {
    IndianaKnox: ['印第安纳州诺克斯', 'yin di an na zhou nuo ke si']
  },
  'America/Matamoros': {
    Matamoros: ['马塔莫罗斯', 'ma ta mo luo si']
  },
  'America/Menominee': {
    Menominee: ['梅诺米尼', 'mei nuo mi ni']
  },
  'America/North_Dakota/New_Salem': {
    NorthDakotaNewSalem: ['北达科他州新塞勒姆', 'bei da ke ta zhou xin sai le mu']
  },
  'America/Rainy_River': {
    RainyRiverDistrict: ['雷尼河', 'lei ni he']
  },
  'America/Rankin_Inlet': {
    RankinInlet: ['兰金因莱特', 'lan jin yin lai te']
  },
  'America/Resolute': {
    Resolute: ['雷索卢特', 'lei suo lu te']
  },
  'America/Indiana/Tell_City': {
    IndianaTellCity: ['印第安纳州特尔城', 'yin di an na zhou te er cheng']
  },
  'America/Winnipeg': {
    Winnipeg: ['温尼伯', 'wen ni bo']
  },
  'America/Eirunepe': {
    Eirunepe: ['埃鲁内佩', 'ai lu nei pei']
  },
  'America/Rio_Branco': {
    RioBranco: ['里奥布朗库', 'li ao bu lang ku']
  },
  'America/Bogota': {
    Bogota: ['波哥大', 'bo ge da'],
    Medellin: ['麦德林', 'mai de lin']
  },
  'Pacific/Easter': {
    EasterIsland: ['复活节岛', 'fu huo jie dao']
  },
  'America/Atikokan': {
    Atikokan: ['阿蒂科肯', 'a di ke ken']
  },
  'America/Cancun': {
    Cancun: ['坎昆', 'kan kun']
  },
  'America/Jamaica': {
    Kingston: ['金斯顿', 'jin si dun']
  },
  'America/Panama': {
    Panama: ['巴拿马', 'ba na ma'],
    PanamaCity: ['巴拿马城', 'ba na ma cheng']
  },
  'America/Guayaquil': {
    Ecuador: ['厄瓜多尔', 'e gua duo er'],
    Quito: ['基多', 'ji duo'],
    Guayaquil: ['瓜亚基尔', 'gua ya ji er']
  },
  'America/Lima': {
    Lima: ['利马', 'li ma']
  },
  'America/Havana': {
    Havana: ['哈瓦那', 'ha wa na']
  },
  'America/Detroit': {
    Detroit: ['底特律', 'di te lv'],
    GrandRapids: ['大急流城', 'da ji liu cheng']
  },
  'America/Grand_Turk': {
    GrandTurk: ['大特克', 'da te ke']
  },
  'America/Indiana/Indianapolis': {
    Indianapolis: ['印第安纳波利斯', 'yin di an na bo li si']
  },
  'America/Iqaluit': {
    Iqaluit: ['伊魁特', 'yi kui te']
  },
  'America/Kentucky/Louisville': {
    Louisville: ['路易斯维尔', 'lu yi si wei er']
  },
  'America/Indiana/Marengo': {
    IndianaMarengo: ['印第安纳州马伦戈', 'yin di an na zhou ma lun ge']
  },
  'America/Kentucky/Monticello': {
    KentuckyMonticello: ['肯塔基州蒙蒂塞洛', 'ken ta ji zhou meng di sai luo']
  },
  'America/Nassau': {
    Nassau: ['拿骚', 'na sao']
  },
  'America/New_York': {
    NewYork: ['纽约', 'niu yue'],
    Washington: ['华盛顿', 'hua sheng dun'],
    Boston: ['波士顿', 'bo shi dun'],
    Miami: ['迈阿密', 'mai a mi'],
    Atlanta: ['亚特兰大', 'ya te lan da'],
    Philadelphia: ['费城', 'fei cheng'],
    Baltimore: ['巴尔的摩', 'ba er di mo'],
    Orlando: ['奥兰多', 'ao lan duo'],
    Columbus: ['哥伦布', 'ge lun bu'],
    Cleveland: ['克利夫兰', 'ke li fu lan'],
    Tampa: ['坦帕', 'tan pa'],
    Cincinnati: ['辛辛那提', 'xin xin na ti'],
    Pittsburgh: ['匹兹堡', 'pi zi bao'],
    VirginiaBeach: ['弗吉尼亚比奇', 'fu ji ni ya bi qi'],
    Charlotte: ['夏洛特', 'xia luo te'],
    Buffalo: ['布法罗', 'bu fa luo'],
    Columbia: ['哥伦比亚', 'ge lun bi ya'],
    Jacksonville: ['杰克逊维尔', 'jie ke xun wei er'],
    Richmond: ['里士满', 'li shi man'],
    Rochester: ['罗彻斯特', 'luo che si te'],
    PalmBay: ['棕榈湾', 'zong lv wan'],
    Ashburn: ['阿什本', 'a shi ben']
  },
  'America/Nipigon': {
    Nipigon: ['尼皮贡', 'ni pi gong']
  },
  'America/Pangnirtung': {
    Pangnirtung: ['庞纳唐', 'pang na tang']
  },
  'America/Indiana/Petersburg': {
    IndianaPetersburg: ['印第安纳州彼得斯堡', 'yin di an na zhou bi de si bao']
  },
  'America/Port-au-Prince': {
    'Port-au-Prince': ['太子港', 'tai zi gang']
  },
  'America/Thunder_Bay': {
    ThunderBay: ['桑德贝', 'sang de bei']
  },
  'America/Toronto': {
    Toronto: ['多伦多', 'duo lun duo'],
    Montreal: ['蒙特利尔', 'meng te li er'],
    Ottawa: ['渥太华', 'wo tai hua'],
    Quebec: ['魁北克', 'kui bei ke'],
    Hamilton: ['汉密尔顿', 'han mi er dun']
  },
  'America/Indiana/Vevay': {
    IndianaVevay: ['印第安纳州维维市', 'yin di an na zhou wei wei shi']
  },
  'America/Indiana/Vincennes': {
    IndianaVincennes: ['印第安纳州温森斯', 'yin di an na zhou wen sen si']
  },
  'America/Indiana/Winamac': {
    IndianaWinamac: ['印第安纳州威纳马克', 'yin di an na zhou wei na ma ke']
  },
  'America/Boa_Vista': {
    BoaVista: ['博阿维斯塔', 'bo a wei si ta']
  },
  'America/Manaus': {
    Manaus: ['马瑙斯', 'ma nao si']
  },
  'America/Porto_Velho': {
    PortoVelho: ['波多韦柳', 'bo duo wei liu']
  },
  'America/Barbados': {
    Bridgetown: ['布里奇顿', 'bu li qi dun']
  },
  'America/Blanc-Sablon': {
    'Blanc-Sablon': ['布兰克萨布隆', 'bu lan ke sa bu long']
  },
  'America/Curacao': {
    Curacao: ['库拉索', 'ku la suo']
  },
  'America/Martinique': {
    Martinique: ['马提尼克', 'ma ti ni ke']
  },
  'America/Port_of_Spain': {
    PortofSpain: ['西班牙港', 'xi ban ya gang']
  },
  'America/Puerto_Rico': {
    PuertoRico: ['波多黎各', 'bo duo li ge']
  },
  'America/Santo_Domingo': {
    SantoDomingo: ['圣多明各', 'sheng duo ming ge']
  },
  'America/La_Paz': {
    Bolivia: ['玻利维亚', 'bo li wei ya']
  },
  'America/Guyana': {
    Guyana: ['圭亚那', 'gui ya na']
  },
  'America/Caracas': {
    Venezuela: ['委内瑞拉', 'wei nei rui la'],
    Caracas: ['加拉加斯', 'jia la jia si']
  },
  'America/Campo_Grande': {
    CampoGrande: ['大坎普', 'da kan pu']
  },
  'America/Cuiaba': {
    Cuiaba: ['库亚巴', 'ku ya ba']
  },
  'Atlantic/Bermuda': {
    Bermuda: ['百慕大', 'bai mu da']
  },
  'America/Glace_Bay': {
    GlaceBay: ['格莱斯贝', 'ge lai si bei']
  },
  'America/Goose_Bay': {
    GooseBay: ['古斯湾', 'gu si wan']
  },
  'America/Halifax': {
    Halifax: ['哈利法克斯', 'ha li fa ke si']
  },
  'America/Moncton': {
    Moncton: ['蒙克顿', 'meng ke dun']
  },
  'America/Thule': {
    Thule: ['图勒', 'tu le']
  },
  'America/Argentina/Buenos_Aires': {
    BuenosAires: ['布宜诺斯艾利斯', 'bu yi nuo si ai li si'],
    PortoAlegre: ['阿雷格里港', 'a lei ge li gang']
  },
  'America/Argentina/Catamarca': {
    Catamarca: ['卡塔马卡', 'ka ta ma ka']
  },
  'America/Argentina/Cordoba': {
    Cordoba: ['科尔多瓦', 'ke er duo wa']
  },
  'America/Argentina/Jujuy': {
    Jujuy: ['胡胡伊', 'hu hu yi']
  },
  'America/Argentina/La_Rioja': {
    LaRioja: ['拉里奥哈', 'la li ao ha']
  },
  'America/Argentina/Mendoza': {
    Mendoza: ['门多萨', 'men duo sa']
  },
  'America/Argentina/Rio_Gallegos': {
    RioGallegos: ['里奥加耶戈斯', 'li ao jia ye ge si']
  },
  'America/Argentina/Salta': {
    Salta: ['萨尔塔', 'sa er ta']
  },
  'America/Argentina/San_Juan': {
    ArgentinaSanJuan: ['圣胡安', 'sheng hu an']
  },
  'America/Argentina/Tucuman': {
    Tucuman: ['图库曼', 'tu ku man']
  },
  'America/Argentina/Ushuaia': {
    Ushuaia: ['乌斯怀亚', 'wu si huai ya']
  },
  'America/Araguaina': {
    Araguaina: ['阿拉瓜伊纳', 'a la gua yi na']
  },
  'America/Bahia': {
    Bahia: ['巴伊亚', 'ba yi ya']
  },
  'America/Belem': {
    Belem: ['贝伦', 'bei lun']
  },
  'America/Fortaleza': {
    Fortaleza: ['福塔雷萨', 'fu ta lei sa']
  },
  'America/Maceio': {
    Maceio: ['马塞约', 'ma sai yue']
  },
  'America/Recife': {
    Recife: ['累西腓', 'lei xi fei']
  },
  'America/Santarem': {
    Santarem: ['圣塔伦', 'sheng ta lun']
  },
  'America/Santiago': {
    Lima: ['利马', 'li ma'],
    Santiago: ['圣地亚哥', 'sheng di ya ge']
  },
  'Atlantic/Stanley': {
    FalklandIslands: ['福克兰群岛', 'fu ke lan qun dao']
  },
  'America/Cayenne': {
    FrenchGuiana: ['法属圭亚那', 'fa shu gui ya na']
  },
  'Antarctica/Palmer': {
    PalmerStation: ['帕默尔站', 'pa mo er zhan']
  },
  'America/Asuncion': {
    Paraguay: ['巴拉圭', 'ba la gui']
  },
  'America/Punta_Arenas': {
    PuntaArenas: ['蓬塔阿雷纳斯', 'peng ta a lei na si']
  },
  'Antarctica/Rothera': {
    RotheraStation: ['罗瑟拉站', 'luo se la zhan']
  },
  'America/Paramaribo': {
    Surinam: ['苏里南', 'su li nan']
  },
  'America/Montevideo': {
    Uruguay: ['乌拉圭', 'wu la gui'],
    Montevideo: ['蒙得维的亚', 'meng de wei de ya']
  },
  'America/Godthab': {
    WestGreenland: ['格陵兰岛西部', 'ge ling lan dao xi bu']
  },
  'America/Argentina/San_Luis': {
    SanLuis: ['圣路易斯', 'sheng lu yi si']
  },
  'America/Sao_Paulo': {
    RioDeJaneiro: ['里约热内卢', 'li yue re nei lu'],
    SaoPaulo: ['圣保罗', 'sheng bao luo'],
    Brasilia: ['巴西利亚', 'ba xi li ya'],
    BeloHorizonte: ['贝洛奥里藏特', 'bei luo ao li cang te']
  },
  'America/St_Johns': {
    Newfoundland: ['纽芬兰', 'niu fen lan']
  },
  'America/Miquelon': {
    StPierreMiquelon: ['圣皮埃尔和密克隆群岛', 'sheng pi ai er he mi ke long qun dao']
  },
  'America/Noronha': {
    FernandoDeNoronha: ['费尔南多-迪诺罗尼亚', 'fei er nan duo - di nuo luo ni ya']
  },
  'Atlantic/South_Georgia': {
    SouthGeorgia: ['南乔治亚岛', 'nan qiao zhi ya dao']
  },
  'Atlantic/Azores': {
    Azores: ['亚速尔群岛', 'ya su er qun dao']
  },
  'Atlantic/Cape_Verde': {
    CapeVerde: ['佛得角', 'fo de jiao']
  },
  'America/Scoresbysund': {
    EastGreenland: ['格陵兰岛东部', 'ge ling lan dao dong bu']
  },
  'Etc/GMT': {
    GreenwichMeanTime: ['格林尼治标准时间', 'ge lin ni zhi biao zhun shi jian']
  },
  'Africa/Abidjan': {
    Abidjan: ['阿比让', 'a bi rang']
  },
  'Africa/Accra': {
    Accra: ['阿克拉', 'a ke la']
  },
  'Africa/Bissau': {
    Bissau: ['比绍', 'bi shao']
  },
  'America/Danmarkshavn': {
    Danmarkshavn: ['丹麦港', 'dan mai gang']
  },
  'Africa/Monrovia': {
    Monrovia: ['蒙罗维亚', 'meng luo wei ya']
  },
  'Atlantic/Reykjavik': {
    Reykjavik: ['雷克雅未克', 'lei ke ya wei ke']
  },
  'Europe/Dublin': {
    Ireland: ['爱尔兰', 'ai er lan'],
    Dublin: ['都柏林', 'dou bo lin']
  },
  'Antarctica/Troll': {
    TrollStation: ['特罗尔站', 'te luo er zhan']
  },
  'Europe/London': {
    London: ['伦敦', 'lun dun'],
    Edinburgh: ['爱丁堡', 'ai ding bao'],
    Birmingham: ['伯明翰', 'bo ming han'],
    Manchester: ['曼彻斯特', 'man che si te'],
    Glasgow: ['格拉斯哥', 'ge la si ge'],
    Liverpool: ['利物浦', 'li wu pu'],
    Nottingham: ['诺丁汉', 'nuo ding han'],
    Coventry: ['考文垂', 'kao wen chui'],
    Southampton: ['南安普顿', 'nan an pu dun'],
    Belfast: ['贝尔法斯特', 'bei er fa si te'],
    MiltonKeynes: ['米尔顿凯恩斯', 'mi er dun kai en si'],
    Cardiff: ['卡迪夫', 'ka di fu'],
    Bradford: ['布拉德福德', 'bu la de fu de'],
    Leeds: ['利兹', 'li zi'],
    Newcastle: ['纽卡斯尔', 'niu ka si er'],
    Sheffield: ['谢菲尔德', 'xie fei er de']
  },
  'Atlantic/Canary': {
    Canary: ['加那利', 'jia na li']
  },
  'Atlantic/Faroe': {
    Faroe: ['法罗', 'fa luo']
  },
  'Europe/Lisbon': {
    Lisbon: ['里斯本', 'li si ben'],
    Porto: ['波尔图', 'bo er tu']
  },
  'Atlantic/Madeira': {
    Madeira: ['马德拉', 'ma de la']
  },
  'Africa/Sao_Tome': {
    SaoTome: ['圣多美', 'sheng duo mei']
  },
  'Africa/Algiers': {
    Algiers: ['阿尔及尔', 'a er ji er'],
    DarElBeida: ['达尔贝达', 'da er bei da']
  },
  'Africa/Tunis': {
    Tunis: ['突尼斯市', 'tu ni si shi']
  },
  'Europe/Amsterdam': {
    Amsterdam: ['阿姆斯特丹', 'a mu si te dan'],
    Rotterdam: ['鹿特丹', 'lu te dan'],
    TheHague: ['海牙', 'hai ya'],
    Utrecht: ['乌得勒支', 'wu de lei zhi'],
    Eindhoven: ['埃因霍温', 'ai yin huo wen']
  },
  'Europe/Andorra': {
    Andorra: ['安道尔', 'an dao er']
  },
  'Europe/Belgrade': {
    Belgrade: ['贝尔格莱德', 'bei er ge lai de'],
    Zagreb: ['萨格勒布', 'sa ge le bu'],
    Ljubljana: ['卢布尔雅那', 'lu bu er ya na']
  },
  'Europe/Berlin': {
    Berlin: ['柏林', 'bo lin'],
    Frankfurt: ['法兰克福', 'fa lan ke fu'],
    Munich: ['慕尼黑', 'mu ni hei'],
    Hamburg: ['汉堡', 'han bao'],
    Stuttgart: ['斯图加特', 'si tu jia te'],
    Dusseldorf: ['杜塞尔多夫', 'du sai er duo fu'],
    Cologne: ['科隆', 'ke long'],
    Essen: ['埃森', 'ai sen'],
    Dresden: ['德累斯顿', 'de lei si dun'],
    Hanover: ['汉诺威', 'han nuo wei'],
    Darmstadt: ['达姆施塔特', 'da mu shi ta te'],
    Karlsruhe: ['卡尔斯鲁厄', 'ka er si lu e'],
    Ingolstadt: ['因戈尔施塔特', 'yin ge er shi ta te'],
    Erlangen: ['埃朗根', 'ai lang gen'],
    Wolfsburg: ['沃尔夫斯堡', 'wo er fu si bao'],
    Regensburg: ['雷根斯堡', 'lei gen si bao'],
    Ulm: ['乌尔姆', 'wu er mu'],
    Nuremberg: ['纽伦堡', 'niu lun bao'],
    Leipzig: ['莱比锡', 'lai bi xi']
  },
  'Europe/Brussels': {
    Brussels: ['布鲁塞尔', 'bu lu sai er'],
    Antwerp: ['安特卫普', 'an te wei pu']
  },
  'Europe/Budapest': {
    Budapest: ['布达佩斯', 'bu da pei si']
  },
  'Africa/Casablanca': {
    Casablanca: ['卡萨布兰卡', 'ka sa bu lan ka']
  },
  'Africa/Ceuta': {
    Ceuta: ['休达', 'xiu da']
  },
  'Europe/Copenhagen': {
    Copenhagen: ['哥本哈根', 'ge ben ha gen'],
    Aarhus: ['奥胡斯', 'ao hu si']
  },
  'Africa/El_Aaiun': {
    ElAaiun: ['阿尤恩', 'a you en']
  },
  'Europe/Gibraltar': {
    Gibraltar: ['直布罗陀', 'zhi bu luo tuo']
  },
  'Europe/Luxembourg': {
    Luxembourg: ['卢森堡', 'lu sen bao']
  },
  'Europe/Madrid': {
    Madrid: ['马德里', 'ma de li'],
    Barcelona: ['巴塞罗那', 'ba sai luo na'],
    Valencia: ['瓦伦西亚', 'wa lun xi ya'],
    Santander: ['桑坦德', 'sang tan de'],
    Seville: ['塞维利亚', 'sai wei li ya'],
    Malaga: ['马拉加', 'ma la jia'],
    ACoruna: ['拉科鲁尼亚', 'la ke lu ni ya']
  },
  'Europe/Malta': {
    Malta: ['马耳他', 'ma er ta']
  },
  'Europe/Monaco': {
    Monaco: ['摩纳哥', 'mo na ge']
  },
  'Europe/Oslo': {
    Oslo: ['奥斯陆', 'ao si lu'],
    Bergen: ['卑尔根', 'bei er gen']
  },
  'Europe/Paris': {
    Paris: ['巴黎', 'ba li'],
    Lyon: ['里昂', 'li ang'],
    Marseille: ['马赛', 'ma sai'],
    Strasbourg: ['斯特拉斯堡', 'si te la si bao'],
    Lille: ['里尔', 'li er'],
    Toulouse: ['图卢兹', 'tu lu zi'],
    Nice: ['尼斯', 'ni si']
  },
  'Europe/Prague': {
    Prague: ['布拉格', 'bu la ge'],
    Bratislava: ['布拉迪斯拉发', 'bu la di si la fa']
  },
  'Europe/Rome': {
    Rome: ['罗马', 'luo ma'],
    Milan: ['米兰', 'mi lan'],
    Venice: ['威尼斯', 'wei ni si'],
    Florence: ['佛罗伦萨', 'fo luo lun sa'],
    Turin: ['都灵', 'du ling'],
    Naples: ['那不勒斯', 'na bu le si'],
    VaticanCity: ['梵蒂冈城', 'fan di gang cheng'],
    Pisa: ['比萨', 'bi sa']
  },
  'Europe/Stockholm': {
    Stockholm: ['斯德哥尔摩', 'si de ge er mo'],
    Malmo: ['马尔默', 'ma er mo'],
    Gothenburg: ['哥德堡', 'ge de bao']
  },
  'Europe/Tirane': {
    Tirane: ['地拉那', 'di la na']
  },
  'Europe/Vienna': {
    Vienna: ['维也纳', 'wei ye na'],
    Linz: ['林茨', 'lin ci']
  },
  'Europe/Warsaw': {
    Warsaw: ['华沙', 'hua sha'],
    Katowice: ['卡托维兹', 'ka tuo wei zi'],
    Tshwane: ['茨瓦内', 'ci wa nei']
  },
  'Europe/Zurich': {
    Zurich: ['苏黎世', 'su li shi'],
    Geneva: ['日内瓦', 'ri nei wa'],
    Lausanne: ['洛桑', 'luo sang']
  },
  'Africa/Lagos': {
    Lagos: ['拉各斯', 'la ge si'],
    Luanda: ['罗安达', 'luo an da'],
    Kinshasa: ['金沙萨', 'jin sha sa']
  },
  'Africa/Ndjamena': {
    Ndjamena: ['恩贾梅纳', 'en jia mei na']
  },
  'Africa/Khartoum': {
    Khartoum: ['喀土穆', 'ka tu mu']
  },
  'Africa/Maputo': {
    Maputo: ['马普托', 'ma pu tuo']
  },
  'Africa/Windhoek': {
    Windhoek: ['温得和克', 'wen de he ke']
  },
  'Africa/Cairo': {
    Cairo: ['开罗', 'kai luo']
  },
  'Europe/Kaliningrad': {
    Kaliningrad: ['加里宁格勒', 'jia li ning ge le']
  },
  'Africa/Tripoli': {
    Tripoli: ['的黎波里', 'di li bo li']
  },
  'Europe/Athens': {
    Athens: ['雅典', 'ya dian']
  },
  'Asia/Beirut': {
    Beirut: ['贝鲁特', 'bei lu te']
  },
  'Europe/Bucharest': {
    Bucharest: ['布加勒斯特', 'bu jia le si te']
  },
  'Europe/Chisinau': {
    Chisinau: ['基希讷乌', 'ji xi ne wu']
  },
  'Asia/Gaza': {
    Gaza: ['加沙', 'jia sha']
  },
  'Asia/Hebron': {
    Hebron: ['希伯伦', 'xi bo lun']
  },
  'Europe/Helsinki': {
    Helsinki: ['赫尔辛基', 'he er xin ji']
  },
  'Europe/Kiev': {
    Kiev: ['基辅', 'ji fu']
  },
  'Asia/Nicosia': {
    Nicosia: ['尼科西亚', 'ni ke xi ya']
  },
  'Europe/Riga': {
    Riga: ['里加', 'li jia']
  },
  'Europe/Sofia': {
    Sofia: ['索菲亚', 'suo fei ya']
  },
  'Europe/Tallinn': {
    Tallinn: ['塔林', 'ta lin']
  },
  'Europe/Uzhgorod': {
    Uzhgorod: ['乌日哥罗德', 'wu ri ge luo de']
  },
  'Europe/Vilnius': {
    Vilnius: ['维尔纽斯', 'wei er niu si']
  },
  'Europe/Zaporozhye': {
    Zaporozhye: ['扎波罗热', 'zha bo luo re']
  },
  'Asia/Famagusta': {
    Famagusta: ['法马古斯塔', 'fa ma gu si ta']
  },
  'Africa/Johannesburg': {
    CapeTown: ['开普敦', 'kai pu dun'],
    Johannesburg: ['约翰内斯堡', 'yue han nei si bao'],
    Durban: ['德班', 'de ban']
  },
  'Africa/Juba': {
    Juba: ['朱巴', 'zhu ba']
  },
  'Asia/Amman': {
    Amman: ['安曼', 'an man']
  },
  'Asia/Damascus': {
    Damascus: ['大马士革', 'da ma shi ge']
  },
  'Asia/Tel_Aviv': {
    TelAviv: ['特拉维夫', 'te la wei fu']
  },
  'Asia/Jerusalem': {
    Jerusalem: ['耶路撒冷', 'ye lu sa leng'],
    Haifa: ['海法', 'hai fa']
  },
  'Asia/Baghdad': {
    Baghdad: ['巴格达', 'ba ge da']
  },
  'Asia/Qatar': {
    Qatar: ['卡塔尔', 'ka ta er'],
    Doha: ['多哈', 'duo ha'],
    Manama: ['麦纳麦', 'mai na mai']
  },
  'Asia/Riyadh': {
    Riyadh: ['利雅得', 'li ya de'],
    Jeddah: ['吉达', 'ji da'],
    KuwaitCity: ['科威特城', 'ke wei te cheng'],
    Dammam: ['达曼', 'da man'],
    Medina: ['麦地那', 'mai de na']
  },
  'Africa/Nairobi': {
    Nairobi: ['内罗毕', 'nei luo bi'],
    Kampala: ['坎帕拉', 'kan pa la']
  },
  'Europe/Kirov': {
    Kirov: ['基洛夫', 'ji luo fu']
  },
  'Europe/Minsk': {
    Minsk: ['明斯克', 'ming si ke']
  },
  'Europe/Moscow': {
    Moscow: ['莫斯科', 'mo si ke'],
    'St.Petersburg': ['圣彼得堡', 'sheng bi de bao'],
    NizhnyNovgorod: ['下诺夫哥罗德', 'xia nuo fu ge luo de'],
    Sochi: ['索契', 'suo qi']
  },
  'Europe/Simferopol': {
    Simferopol: ['辛菲罗波尔', 'xin fei luo bo er']
  },
  'Antarctica/Syowa': {
    SyowaStation: ['昭和站', 'zhao he zhan']
  },
  'Europe/Istanbul': {
    Istanbul: ['伊斯坦布尔', 'yi si tan bu er'],
    Ankara: ['安卡拉', 'an ka la'],
    Izmir: ['伊兹密尔', 'yi zi mi er']
  },
  'Europe/Volgograd': {
    Volgograd: ['伏尔加格勒', 'fu er jia ge le']
  },
  'Asia/Yerevan': {
    Yerevan: ['埃里温', 'ai li wen']
  },
  'Europe/Astrakhan': {
    Astrakhan: ['阿斯特拉罕', 'a si te la han']
  },
  'Asia/Baku': {
    Baku: ['巴库', 'ba ku']
  },
  'Asia/Tbilisi': {
    Tbilisi: ['第比利斯', 'di bi li si']
  },
  'Asia/Dubai': {
    Dubai: ['迪拜', 'di bai'],
    AbuDhabi: ['阿布扎比', 'a bu zha bi'],
    Muscat: ['马斯喀特', 'ma si ka te']
  },
  'Indian/Mauritius': {
    Mauritius: ['毛里求斯', 'mao li qiu si']
  },
  'Indian/Reunion': {
    Reunion: ['留尼汪', 'liu ni wang']
  },
  'Europe/Samara': {
    Samara: ['萨马拉', 'sa ma la']
  },
  'Europe/Saratov': {
    Saratov: ['萨拉托夫', 'sa la tuo fu']
  },
  'Indian/Mahe': {
    Seychelles: ['塞舌尔', 'sai she er']
  },
  'Europe/Ulyanovsk': {
    Ulyanovsk: ['乌里扬诺夫斯克', 'wu li yang nuo fu si ke']
  },
  'Asia/Tehran': {
    Tehran: ['德黑兰', 'de hei lan']
  },
  'Asia/Kabul': {
    Kabul: ['喀布尔', 'ka bu er']
  },
  'Indian/Kerguelen': {
    KerguelenIslands: ['凯尔盖朗群岛', 'kai er gai lang qun dao']
  },
  'Indian/Maldives': {
    Maldives: ['马尔代夫', 'ma er dai fu']
  },
  'Antarctica/Mawson': {
    MawsonStation: ['莫森站', 'mo sen zhan']
  },
  'Asia/Karachi': {
    Islamabad: ['伊斯兰堡', 'yi si lan bao'],
    Karachi: ['卡拉奇', 'ka la qi'],
    Lahore: ['拉合尔', 'la he er']
  },
  'Asia/Dushanbe': {
    Dushanbe: ['杜尚别', 'du shang bie']
  },
  'Asia/Ashgabat': {
    Ashgabat: ['阿什喀巴德', 'a shi ka ba de']
  },
  'Asia/Samarkand': {
    Samarkand: ['撒马尔罕', 'sa ma er han']
  },
  'Asia/Tashkent': {
    Tashkent: ['塔什干', 'ta shi gan']
  },
  'Asia/Aqtau': {
    Aqtau: ['阿克套', 'a ke tao']
  },
  'Asia/Aqtobe': {
    Aqtobe: ['阿克托别', 'a ke tuo bie']
  },
  'Asia/Atyrau': {
    Atyrau: ['阿特劳', 'a te lao']
  },
  'Asia/Oral': {
    Oral: ['乌拉尔', 'wu la er']
  },
  'Asia/Yekaterinburg': {
    Yekaterinburg: ['叶卡捷琳堡', 'ye ka jie lin bao']
  },
  'Asia/Qyzylorda': {
    Qyzylorda: ['克孜洛尔达', 'ke zi luo er da']
  },
  'Asia/Colombo': {
    Colombo: ['科伦坡', 'ke lun po']
  },
  'Asia/Calcutta': {
    Calcutta: ['加尔各答', 'jia er ge da'],
    Mumbai: ['孟买', 'meng mai'],
    NewDelhi: ['新德里', 'xin de li'],
    Bangalore: ['班加罗尔', 'ban jia luo er'],
    Hyderabad: ['海德拉巴', 'hai de la ba'],
    Chennai: ['金奈', 'jin nai'],
    Ahmedabad: ['艾哈迈达巴德', 'ai ha mai da ba de'],
    Pune: ['浦那', 'pu na'],
    Surat: ['苏拉特', 'su la te'],
    Gurgaon: ['古爾岡', 'gu er gang']
  },
  'Asia/Katmandu': {
    Kathmandu: ['加德满都', 'jia de man du']
  },
  'Asia/Dhaka': {
    Dhaka: ['达卡', 'da ka']
  },
  'Asia/Thimphu': {
    Thimphu: ['廷布', 'ting bu']
  },
  'Asia/Almaty': {
    Almaty: ['阿拉木图', 'a la mu tu']
  },
  'Indian/Chagos': {
    ChagosArchipelago: ['查戈斯群岛', 'cha ge si qun dao']
  },
  'Asia/Bishkek': {
    Bishkek: ['比什凯克', 'bi shi kai ke']
  },
  'Asia/Omsk': {
    Omsk: ['鄂木斯克', 'e mu si ke']
  },
  'Asia/Urumqi': {
    Urumqi: ['乌鲁木齐', 'wu lu mu qi']
  },
  'Antarctica/Vostok': {
    Vostok: ['沃斯托克', 'wo si tuo ke']
  },
  'Indian/Cocos': {
    CocosIslands: ['科科斯群岛', 'ke ke si qun dao']
  },
  'Asia/Yangon': {
    Yangon: ['仰光', 'yang guang']
  },
  'Asia/Barnaul': {
    Barnaul: ['巴尔瑙尔', 'ba er nao er']
  },
  'Indian/Christmas': {
    ChristmasIsland: ['圣诞岛', 'sheng dan dao']
  },
  'Antarctica/Davis': {
    DavisStation: ['戴维斯站', 'dai wei si zhan']
  },
  'Asia/Hovd': {
    Hovd: ['科布多', 'ke bu duo']
  },
  'Asia/Bangkok': {
    Bangkok: ['曼谷', 'man gu'],
    ChiangMai: ['清迈', 'qing mai'],
    PhnomPenh: ['金边', 'jin bian']
  },
  'Asia/Saigon': {
    HoChiMinhCity: ['胡志明市', 'hu zhi ming shi'],
    Hanoi: ['河内', 'he nei']
  },
  'Asia/Krasnoyarsk': {
    Krasnoyarsk: ['克拉斯诺亚尔斯克', 'ke la si nuo ya er si ke']
  },
  'Asia/Novokuznetsk': {
    Novokuznetsk: ['新库兹涅茨克', 'xin ku zi nie ci ke']
  },
  'Asia/Novosibirsk': {
    Novosibirsk: ['新西伯利亚', 'xin xi bo li ya']
  },
  'Asia/Tomsk': {
    Tomsk: ['托木斯克', 'tuo mu si ke']
  },
  'Asia/Jakarta': {
    Jakarta: ['雅加达', 'ya jia da'],
    Surabaya: ['泗水', 'si shui'],
    Bandung: ['万隆', 'wan long']
  },
  'Asia/Pontianak': {
    Pontianak: ['坤甸', 'kun dian']
  },
  'Australia/Perth': {
    Perth: ['珀斯', 'po si']
  },
  'Asia/Brunei': {
    BandarSeriBegawan: ['斯里巴加湾', 'si li ba jia wan']
  },
  'Asia/Makassar': {
    Makassar: ['望加锡', 'wang jia xi']
  },
  'Asia/Macau': {
    Macao: ['澳门', 'ao men']
  },
  'Asia/Shanghai': {
    Beijing: ['北京', 'bei jing'],
    Chongqing: ['重庆', 'chong qing'],
    Shanghai: ['上海', 'shang hai'],
    Tianjin: ['天津', 'tian jin'],
    Jinan: ['济南', 'ji nan'],
    Qingdao: ['青岛', 'qing dao'],
    Yantai: ['烟台', 'yan tai'],
    Zhengzhou: ['郑州', 'zheng zhou'],
    Taiyuan: ['太原', 'tai yuan'],
    Linfen: ['临汾', 'lin fen'],
    Shenyang: ['沈阳', 'shen yang'],
    Nanjing: ['南京', 'nan jing'],
    Wuxi: ['无锡', 'wu xi'],
    Suzhou: ['苏州', 'su zhou'],
    Hangzhou: ['杭州', 'hang zhou'],
    Ningbo: ['宁波', 'ning bo'],
    Yiwu: ['义乌', 'yi wu'],
    Wenzhou: ['温州', 'wen zhou'],
    Wuhan: ['武汉', 'wu han'],
    Hefei: ['合肥', 'he fei'],
    Nanchang: ['南昌', 'nan chang'],
    Chengdu: ['成都', 'cheng du'],
    Neijiang: ['内江', 'nei jiang'],
    Yibin: ['宜宾', 'yi bin'],
    Xian: ['西安', 'xi an'],
    Lanzhou: ['兰州', 'lan zhou'],
    Kunming: ['昆明', 'kun ming'],
    Zhuhai: ['珠海', 'zhu hai'],
    Shenzhen: ['深圳', 'shen zhen'],
    Dongguan: ['东莞', 'dong guan'],
    Foshan: ['佛山', 'fo shan'],
    Guangzhou: ['广州', 'guang zhou'],
    Qingyuan: ['清远', 'qing yuan'],
    Fuzhou: ['福州', 'fu zhou'],
    Xiamen: ['厦门', 'xia men'],
    Longyan: ['龙岩', 'long yan'],
    Changsha: ['长沙', 'chang sha'],
    Haikou: ['海口', 'hai kou'],
    Sanya: ['三亚', 'san ya'],
    Chengmai: ['澄迈', 'cheng mai']
  },
  'Asia/Choibalsan': {
    Choibalsan: ['乔巴山', 'qiao ba shan']
  },
  'Asia/Hong_Kong': {
    Hongkong: ['香港', 'xiang gang']
  },
  'Asia/Irkutsk': {
    Irkutsk: ['伊尔库茨克', 'yi er ku ci ke']
  },
  'Asia/Kuala_Lumpur': {
    KualaLumpur: ['吉隆坡', 'ji long po'],
    Chittagong: ['吉大港', 'ji da gang']
  },
  'Asia/Kuching': {
    Kuching: ['古晋', 'gu jin']
  },
  'Asia/Manila': {
    Manila: ['马尼拉', 'ma ni la']
  },
  'Asia/Singapore': {
    Singapore: ['新加坡', 'xin jia po']
  },
  'Asia/Taipei': {
    Taipei: ['台北', 'tai bei']
  },
  'Asia/Ulaanbaatar': {
    Ulaanbaatar: ['乌兰巴托', 'wu lan ba tuo']
  },
  'Australia/Eucla': {
    Eucla: ['尤克拉', 'you ke la']
  },
  'Asia/Dili': {
    Dili: ['帝力', 'di li']
  },
  'Asia/Jayapura': {
    Jayapura: ['查亚普拉', 'cha ya pu la']
  },
  'Asia/Tokyo': {
    Tokyo: ['东京', 'dong jing'],
    Osaka: ['大阪', 'da ban'],
    Nagoya: ['名古屋', 'ming gu wu'],
    Yokohama: ['横滨', 'heng bin'],
    Kyoto: ['京都', 'jing du'],
    Fukuoka: ['福冈', 'fu gang'],
    Sapporo: ['札幌', 'zha huang'],
    Kobe: ['神户', 'shen hu'],
    Saitama: ['埼玉', 'qi yu'],
    Okinawa: ['冲绳', 'chong sheng'],
    Fuji: ['富士', 'fu shi'],
    Hiroshima: ['广岛', 'guang dao'],
    Nagasaki: ['长崎', 'chang qi'],
    Chiba: ['千叶', 'qian ye'],
    Sendai: ['仙台', 'xian tai'],
    Kawasaki: ['川崎', 'chuan qi'],
    Okayama: ['冈山', 'gang shan'],
    Nagano: ['长野', 'chang ye']
  },
  'Asia/Pyongyang': {
    Pyongyang: ['平壤', 'ping rang']
  },
  'Asia/Seoul': {
    Seoul: ['首尔', 'shou er'],
    Busan: ['釜山', 'fu shan'],
    Incheon: ['仁川', 'ren chuan'],
    Ulsan: ['蔚山', 'wei shan'],
    Changwon: ['昌原', 'chang yuan'],
    Daegu: ['大邱', 'da qiu']
  },
  'Pacific/Palau': {
    Palau: ['帕劳', 'pa lao']
  },
  'Asia/Chita': {
    Chita: ['赤塔', 'chi ta']
  },
  'Asia/Khandyga': {
    Khandyga: ['汉德加', 'han de jia']
  },
  'Asia/Yakutsk': {
    Yakutsk: ['雅库茨克', 'ya ku ci ke']
  },
  'Australia/Darwin': {
    Darwin: ['达尔文', 'da er wen']
  },
  'Australia/Brisbane': {
    Brisbane: ['布里斯班', 'bu li si ban'],
    GoldCoast: ['黄金海岸', 'huang jin hai an']
  },
  'Australia/Lindeman': {
    LindemanIsland: ['林德曼', 'lin de man']
  },
  'Pacific/Guam': {
    Chamorro: ['查莫罗', 'cha mo luo']
  },
  'Pacific/Chuuk': {
    Chuuk: ['楚克', 'chu ke']
  },
  'Antarctica/DumontDUrville': {
    DumontDUrvilleStation: ['迪蒙迪尔维尔站', 'di meng di er wei er zhan']
  },
  'Pacific/Port_Moresby': {
    PortMoresby: ['莫尔斯比港', 'mo er si bi gang']
  },
  'Asia/Ust-Nera': {
    'Ust-Nera': ['乌斯内拉', 'wu si nei la']
  },
  'Asia/Vladivostok': {
    Vladivostok: ['符拉迪沃斯托克', 'fu la di wo si tuo ke']
  },
  'Australia/Adelaide': {
    Adelaide: ['阿德莱德', 'a de lai de']
  },
  'Australia/Broken_Hill': {
    BrokenHill: ['布罗肯希尔', 'bu luo ken xi er']
  },
  'Antarctica/Casey': {
    CaseyStation: ['凯西站', 'kai xi zhan']
  },
  'Pacific/Bougainville': {
    Bougainville: ['布干维尔', 'bu gan wei er']
  },
  'Australia/Currie': {
    Currie: ['库利', 'ku li']
  },
  'Australia/Hobart': {
    Hobart: ['霍巴特', 'huo ba te']
  },
  'Australia/Melbourne': {
    Melbourne: ['墨尔本', 'mo er ben']
  },
  'Australia/Sydney': {
    Sydney: ['悉尼', 'xi ni'],
    Canberra: ['堪培拉', 'kan pei la']
  },
  'Pacific/Kosrae': {
    Kosrae: ['科斯雷', 'ke si lei']
  },
  'Australia/Lord_Howe': {
    LordHoweIsland: ['豪勋爵岛', 'hao xun jue dao']
  },
  'Antarctica/Macquarie': {
    MacquarieIsland: ['麦夸里岛', 'mai kua li dao']
  },
  'Asia/Magadan': {
    Magadan: ['马加丹', 'ma jia dan']
  },
  'Pacific/Noumea': {
    NewCaledonia: ['新喀里多尼亚', 'xin ka li duo ni ya']
  },
  'Pacific/Pohnpei': {
    Pohnpei: ['波纳佩', 'bo na pei']
  },
  'Asia/Sakhalin': {
    Sakhalin: ['库页岛', 'ku ye dao']
  },
  'Pacific/Guadalcanal': {
    SolomonIslands: ['所罗门群岛', 'suo luo men qun dao']
  },
  'Asia/Srednekolymsk': {
    Srednekolymsk: ['中科雷姆斯克', 'zhong ke lei mu si ke']
  },
  'Pacific/Efate': {
    Efate: ['埃法特岛', 'ai fa te dao']
  },
  'Pacific/Norfolk': {
    NorfolkIsland: ['诺福克岛', 'nuo fu ke dao']
  },
  'Asia/Anadyr': {
    Anadyr: ['阿纳德尔', 'a na de er']
  },
  'Pacific/Tarawa': {
    GilbertIslands: ['吉尔伯特群岛', 'ji er bo te qun dao']
  },
  'Pacific/Kwajalein': {
    Kwajalein: ['夸贾林', 'kua jia lin']
  },
  'Pacific/Majuro': {
    Majuro: ['马朱罗', 'ma zhu luo']
  },
  'Pacific/Nauru': {
    Nauru: ['瑙鲁', 'nao lu']
  },
  'Asia/Kamchatka': {
    PetropavlovskKamchatski: ['堪察加彼得巴甫洛夫斯克', 'kan cha jia bi de ba fu luo fu si ke']
  },
  'Pacific/Funafuti': {
    Tuvalu: ['图瓦卢', 'tu wa lu']
  },
  'Pacific/Wake': {
    WakeIsland: ['威克岛', 'wei ke dao']
  },
  'Pacific/Wallis': {
    MataUtu: ['马塔乌图', 'ma ta wu tu']
  },
  'Pacific/Fiji': {
    Fiji: ['斐济', 'fei ji']
  },
  'Pacific/Auckland': {
    Wellington: ['惠灵顿', 'hui ling dun'],
    Auckland: ['奥克兰', 'ao ke lan'],
    Christchurch: ['基督城', 'ji du cheng']
  },
  'Pacific/Enderbury': {
    PhoenixIslands: ['菲尼克斯群岛', 'fei ni ke si qun dao']
  },
  'Pacific/Fakaofo': {
    Tokelau: ['托克劳', 'tuo ke lao']
  },
  'Pacific/Tongatapu': {
    NukuAlofa: ['努库阿洛法', 'nu ku a luo fa']
  },
  'Pacific/Apia': {
    Apia: ['阿皮亚', 'a pi ya']
  },
  'Pacific/Chatham': {
    ChathamIslands: ['查塔姆群岛', 'cha ta mu qun dao']
  },
  'Pacific/Kiritimati': {
    LineIslands: ['莱恩群岛', 'lai en qun dao']
  }
};
