// @universe-design/react/lib/pro/timezone-picker/utils.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.searchTimezoneMetaList = exports.importTimezoneLocale = exports.getTimezoneNames = exports.getTimezoneMeta = exports.getTimezoneFormatData = exports.fillInTimezoneMetaList = exports.defaultFormatter = exports.chinaMainlandMetaTransformer = exports.arrayJoin = void 0;
exports.splitKeywords = splitKeywords;
exports.useTimezoneLocale = exports.useMetaTransformer = void 0;
var _interopRequireWildcard2 = _interopRequireDefault(require("@babel/runtime/helpers/interopRequireWildcard"));
var _momentTimezone = require("@lark/moment-timezone");
var _compact = _interopRequireDefault(require("lodash/compact"));
var _countBy = _interopRequireDefault(require("lodash/countBy"));
var _isEmpty = _interopRequireDefault(require("lodash/isEmpty"));
var _keys = _interopRequireDefault(require("lodash/keys"));
var _memoize = _interopRequireDefault(require("lodash/memoize"));
var _useContextSelector = require("use-context-selector");
var _const = require("./const");
var _configProvider = require("../../config-provider");
var _useImportLocale = require("../../config-provider/hooks/useImportLocale");
var _enUS = _interopRequireDefault(require("./locales/en-US"));
// 提取gmt偏移量
const GMT_STRING_REGEXP = /\(?(gmt|utc)?\s*(\+?|-)\s*([\d:]+)([^\d:])*/i;
function normalizeGMTOffset() {
  let gmtString = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
  return gmtString.replace(GMT_STRING_REGEXP, ($0, $1, $2, offset) => $2 + offset.replace(/(\d{2})/g, '$1:').replace(/(:\B)/g, ''));
}

/**
 * 包含 compatible 逻辑的获取 timezone meta 函数
 * @param value - timezone 值
 * @returns timezone meta值
 */
const getTimezoneMeta = value => {
  if (!value) {
    return void 0;
  }
  const timezoneMetaMap = (0, _const.getTimezoneMetaMap)();
  return timezoneMetaMap[value] || timezoneMetaMap[_const.TIMEZONE_LINK_MAP[value]];
};
exports.getTimezoneMeta = getTimezoneMeta;
// 预期搜索 “+8” 也能匹配 “+08:00” 的选项
function splitKeywords(keywords) {
  return keywords.replace(/\+(\d)(?!\d)/g, '+ $1').split(/\s+/g);
}
const searchTimezoneMetaList = (timezoneMetaList, keywords) => {
  if (!keywords) {
    return timezoneMetaList;
  }
  return (0, _compact.default)(timezoneMetaList.map(item => {
    const _text = item.l.toLowerCase();
    const _keywords = keywords.toLowerCase();
    const matchedSearchEntries = item.s.filter(entry => {
      var _entry$c;
      const [enPure, zhCnPure, zhCnPinyinFull, zhCnPinyinInitials = '', zhCnPinyinIndexes = []] = entry;
      if ((_entry$c = entry.c) !== null && _entry$c !== void 0 && _entry$c.toLowerCase().includes(_keywords)) {
        return true;
      }
      if (zhCnPure.includes(keywords)) {
        return true;
      }
      if (enPure.toLowerCase().includes(_keywords)) {
        return true;
      }
      if (zhCnPinyinInitials.includes(keywords)) {
        return true;
      }
      if (zhCnPinyinIndexes.includes(zhCnPinyinFull.replace(/\s/g, '').indexOf(keywords))) {
        return true;
      }
      return false;
    });
    if (matchedSearchEntries.length > 0) {
      return {
        ...item,
        cities: matchedSearchEntries.map(entry => entry.c)
      };
    }
    if (_text.includes(_keywords)) {
      return item;
    }
    const words = splitKeywords(_keywords);
    if (words.every(w => _text.includes(w))) {
      return item;
    }
    const gmtOffset = normalizeGMTOffset(_text);
    if (gmtOffset.includes(normalizeGMTOffset(keywords))) {
      return item;
    }
    return null;
  }, []));
};
exports.searchTimezoneMetaList = searchTimezoneMetaList;
const extendSearchEntries = (searchEntries, localeData) => searchEntries.map(entry => {
  // order: [enPure, zhCnPure, zhCnPinyinFull]
  const zhCnPinyinFull = entry[2];
  const zhCnPinyinFullList = zhCnPinyinFull.split(/\s+/g);
  const zhCnPinyinInitials = zhCnPinyinFullList.map(w => w[0]).join('');
  const zhCnPinyinIndexes = zhCnPinyinFullList.reduce((res, curr) => {
    res.push(res[res.length - 1] + curr.length);
    return res;
  }, [0]);
  const extendedEntry = Object.assign([entry[0], entry[1], entry[2], zhCnPinyinInitials, zhCnPinyinIndexes], {
    c: localeData[entry.c]
  });
  return extendedEntry;
});

/**
 * 默认格式化函数
 */
const defaultFormatter = p => p ? `(${p.GMTOffset}) ${p.zoneName}` : '';
exports.defaultFormatter = defaultFormatter;
const getTimezoneNameCount = (0, _memoize.default)(localeData => (0, _countBy.default)(_const.TIMEZONE_META_LIST, item => localeData[item.l]));
getTimezoneNameCount.cache = new WeakMap();

/**
 * timezone 格式化
 * @internal
 * @experimental
 * @param value - timezone 值
 * @param localeData - timezone 国际化文案数据
 * @returns 格式化数据
 */
const getTimezoneFormatData = function (value, localeData) {
  var _options$transformer;
  let options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
  const meta = getTimezoneMeta(value);
  if (!meta) {
    return void 0;
  }
  const localeKey = ((_options$transformer = options.transformer) === null || _options$transformer === void 0 ? void 0 : _options$transformer.call(options, meta, (0, _const.getTimezoneMetaMap)()).l) || meta.l;
  const timeName = localeData[localeKey];
  const isUnique = getTimezoneNameCount(localeData)[timeName] === 1;
  const cityLocales = (0, _keys.default)(options.cities).map(city => localeData[`Ct_${city}`]);
  return {
    GMTOffset: (0, _momentTimezone.getGMTTimezone)(value, {
      padNumber: true,
      date: options.date
    }),
    timeName,
    zoneName: `${timeName}${isUnique || (0, _isEmpty.default)(cityLocales) ? '' : ` - ${cityLocales[0]}`}`,
    isUnique,
    cities: cityLocales
  };
};
exports.getTimezoneFormatData = getTimezoneFormatData;
const fillInTimezoneMetaList = (localeData, date, transformer) => {
  const _sortedTimezoneMetaList = (0, _const.getSortedTimezoneMetaList)();
  const _timezoneList = _sortedTimezoneMetaList.map(item => {
    const {
      v: value,
      c: cities,
      s: searchEntries
    } = item;
    const temp = {
      v: value,
      c: cities,
      l: defaultFormatter(getTimezoneFormatData(value, localeData, {
        date,
        cities,
        transformer
      })),
      s: extendSearchEntries(searchEntries, localeData)
    };
    return temp;
  });
  return _timezoneList;
};

/**
 * @public
 * 获取所有时区的名称值
 */
exports.fillInTimezoneMetaList = fillInTimezoneMetaList;
const getTimezoneNames = () => _const.TIMEZONE_META_LIST.map(o => o.v);
exports.getTimezoneNames = getTimezoneNames;
const arrayJoin = (array, joiner) => {
  if (!array) {
    return [];
  }
  const result = [];
  for (let i = 0; i < array.length; ++i) {
    result.push(array[i]);
    if (i !== array.length - 1) {
      result.push(joiner(i));
    }
  }
  return result;
};
exports.arrayJoin = arrayJoin;
const importTimezoneLocale = async locale => {
  let promise = (specifier => new Promise(r => r(specifier)).then(s => (0, _interopRequireWildcard2.default)(require(s))))(`./locales/${locale}.js`);
  return promise.then(_ref => {
    let {
      default: localeData
    } = _ref;
    return localeData;
  });
};
exports.importTimezoneLocale = importTimezoneLocale;
const useTimezoneLocale = () => (0, _useImportLocale.useImportLocale)({
  importLocale: importTimezoneLocale,
  fallbackLocale: _enUS.default
});
exports.useTimezoneLocale = useTimezoneLocale;
const useMetaTransformer = () => (0, _useContextSelector.useContextSelector)(_configProvider.ConfigContext, o => {
  var _o$timezone;
  return (_o$timezone = o.timezone) === null || _o$timezone === void 0 ? void 0 : _o$timezone.metaTransformer;
});
exports.useMetaTransformer = useMetaTransformer;
const chinaMainlandMetaTransformer = (meta, metaMap) => {
  if (meta.v === 'Asia/Hong_Kong' || meta.v === 'Asia/Macau' || meta.v === 'Asia/Taipei') {
    return {
      ...meta,
      l: metaMap['Asia/Shanghai'].l
    };
  }
  return meta;
};
exports.chinaMainlandMetaTransformer = chinaMainlandMetaTransformer;
