// @universe-design/react/es/upload/uploadFile/CardFile.js
//.u1jhi6b8{flex-grow:0;}
import ug24l6d from "./../../.stylex/ug24l6d";
//.ui64zpg{flex-shrink:0;}
import uidkx17 from "./../../.stylex/uidkx17";
//.u1euv43f{position:absolute;}
import u1u1p7mj from "./../../.stylex/u1u1p7mj";
//.u5gq2j6{bottom:0px;}
import u8rtfz1 from "./../../.stylex/u8rtfz1";
//.u9j3dip{left:8px;}
import ui763hu from "./../../.stylex/ui763hu";
//.u112f5iw{right:8px;}
import ubo3qzx from "./../../.stylex/ubo3qzx";
//.u10pi13n{position:relative;}
import u5jxv71 from "./../../.stylex/u5jxv71";
import clsx from 'clsx';
import React from 'react';
import stylex from '@universe-design/stylex';
import { UploadProgress } from './UploadProgress';
import { isImage } from '../utils';
import { FileContextProvider } from './FileContext';
import { useFile } from './useFile';
import { FileName } from './FileName';
import '../style/card-file.less';
import { devWarning, warning } from '../../util';
export const CardFile = fileProps => {
  const {
    style,
    nameRenderer
  } = fileProps;
  const fileCtxValue = useFile(fileProps);
  const {
    classNames,
    file,
    errorNode,
    retryNode,
    removeNode,
    downloadNode,
    previewNode,
    iconRender,
    rootPrefixCls,
    progressProps
  } = fileCtxValue;
  process.env.NODE_ENV !== "production" ? warning(!['downloading', 'downloadFailed', 'downloaded'].includes(file.status), devWarning('Upload', `file.status=${file.status} does not support for CardFile.`)) : void 0;
  return /*#__PURE__*/React.createElement(FileContextProvider, {
    value: fileCtxValue
  }, /*#__PURE__*/React.createElement("div", {
    className: clsx(classNames, `${rootPrefixCls}--card`),
    style: style
  }, /*#__PURE__*/React.createElement("div", {
    className: clsx(`${rootPrefixCls}__preview`, stylex({
      "props": {
        "qhf8xq": u5jxv71
      }
    }))
  }, function () {
    const previewFile = file;
    const previewUrl = previewFile.thumbUrl || previewFile.url;
    if (isImage(previewFile) && previewUrl && previewFile.status === 'success') {
      return /*#__PURE__*/React.createElement("div", {
        className: `${rootPrefixCls}-preview-img`
      }, /*#__PURE__*/React.createElement("img", {
        src: previewUrl
      }));
    }
    return /*#__PURE__*/React.createElement("div", {
      className: `${rootPrefixCls}-preview-icon`
    }, iconRender(previewFile));
  }(), /*#__PURE__*/React.createElement(UploadProgress, {
    file: file,
    progressProps: progressProps,
    prefixCls: rootPrefixCls,
    className: stylex({
      "props": {
        "qhf8xq": u1u1p7mj,
        "B5kzvoi": u8rtfz1,
        "oyh7mz": ui763hu,
        "j35jbq": ubo3qzx
      }
    })
  })), /*#__PURE__*/React.createElement("div", {
    className: `${rootPrefixCls}--card__info`
  }, /*#__PURE__*/React.createElement("span", {
    className: `${rootPrefixCls}--card__info-icon`
  }, iconRender(file)), /*#__PURE__*/React.createElement("div", {
    className: `${rootPrefixCls}--card__info-name`
  }, /*#__PURE__*/React.createElement(FileName, {
    file: file,
    nameRenderer: nameRenderer
  })), /*#__PURE__*/React.createElement("span", {
    className: clsx(`${rootPrefixCls}--card__info-operations`, stylex({
      "props": {
        "Bh6795r": ug24l6d,
        "Bnnss6s": uidkx17
      }
    }))
  }, errorNode, previewNode, downloadNode, retryNode, removeNode))));
};
