// @universe-design/react/es/upload/uploadFile/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import * as React from 'react';
import { CardFile } from './CardFile';
import { PictureFile } from './PictureFile';
import { TextFile, TextFileAction } from './TextFile';
import { useUploadFile } from './useUploadFile';
export const UploadTextFile = props => {
  const uploadFileProps = useUploadFile(props);
  return /*#__PURE__*/React.createElement(TextFile, uploadFileProps);
};
const SimpleTextFile = props => /*#__PURE__*/React.createElement(TextFile, _extends({}, props, {
  isSimple: true
}));
export const UploadSimpleTextFile = props => {
  const uploadFileProps = useUploadFile(props);
  return /*#__PURE__*/React.createElement(SimpleTextFile, uploadFileProps);
};
export const UploadCardFile = props => {
  const uploadFileProps = useUploadFile(props);
  return /*#__PURE__*/React.createElement(CardFile, uploadFileProps);
};
export const UploadPictureFile = props => {
  const uploadFileProps = useUploadFile(props);
  return /*#__PURE__*/React.createElement(PictureFile, uploadFileProps);
};
export { TextFile, TextFileAction, CardFile, PictureFile };
