// @universe-design/react/es/upload/uploadFile/TextFile.js
//.u1k6fduh{cursor:pointer;}
import u12o6wmt from "./../../.stylex/u12o6wmt";
/** @jsxRuntime classic */
/** @jsx jsx */
/** @jsxFrag React.Fragment */
import clsx from 'clsx';
import React from 'react';
import { useLocaleReceiver } from '../../config-provider';
import { TextOverflow } from '../../shared/components/text-overflow';
import { UploadProgress } from './UploadProgress';
import { formatFileSize, isError } from '../utils';
import { useFile } from './useFile';
import { FileName } from './FileName';
import { FileContextProvider, useFileContext } from './FileContext';
import stylex from '@universe-design/stylex';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx } from '@universe-design/stylex/react';
import '../style/text-file.less';
const DownloadAction = () => {
  const {
    downloadNode
  } = useFileContext();
  return downloadNode;
};
const RetryAction = () => {
  const {
    retryNode
  } = useFileContext();
  return retryNode;
};
const PreviewAction = () => {
  const {
    previewNode
  } = useFileContext();
  return previewNode;
};
const RemoveAction = () => {
  const {
    removeNode
  } = useFileContext();
  return removeNode;
};
export const TextFileAction = {
  Download: DownloadAction,
  Retry: RetryAction,
  Preview: PreviewAction,
  Remove: RemoveAction
};

/**
 * 与 Upload 无关的 File 组件
 */

const FileStatus = _ref => {
  let {
    file,
    prefixCls
  } = _ref;
  const locale = useLocaleReceiver('upload');
  if (file.status === 'waiting') {
    return jsx(TextOverflow, {
      as: "p",
      className: `${prefixCls}-info-subtitle`,
      lineClamp: 1
    }, locale.waiting);
  }
  if (isError(file)) {
    const defaultFailedReason = file.status === 'error' ? locale.failed : locale.downloadFailed;
    const failedReason = file.failedReason || defaultFailedReason;
    return jsx(TextOverflow, {
      as: "p",
      className: `${prefixCls}-info-error`,
      lineClamp: 2
    }, failedReason);
  }
  if (file.status === 'uploading' || file.status === 'downloading') {
    return jsx(TextOverflow, {
      as: "p",
      className: `${prefixCls}-info-subtitle`,
      lineClamp: 1
    }, file.status === 'downloading' && `${locale.downloading} `, formatFileSize(file.loaded, locale), " / ", formatFileSize(file.size, locale));
  }
  return jsx(TextOverflow, {
    as: "p",
    className: `${prefixCls}-info-subtitle`,
    lineClamp: 1
  }, formatFileSize(file.size, locale));
};
const defaultAction = jsx(React.Fragment, null, jsx(TextFileAction.Preview, {
  key: "preview"
}), jsx(TextFileAction.Retry, {
  key: "retry"
}), jsx(TextFileAction.Download, {
  key: "download"
}), jsx(TextFileAction.Remove, {
  key: "remove"
}));
export function TextFile(props) {
  const {
    style,
    isSimple,
    nameRenderer,
    action = defaultAction
  } = props;
  const fileCtxValue = useFile(props);
  const {
    file,
    errorNode,
    iconRender,
    rootPrefixCls,
    onPreview,
    progressProps,
    classNames
  } = fileCtxValue;
  const operationAreaRef = React.useRef(null);
  const previewable = ['success', 'downloading', 'downloaded', 'downloadFailed'].includes(file.status);
  return jsx(FileContextProvider, {
    value: fileCtxValue
  }, jsx("div", {
    className: clsx(classNames, isSimple ? `${rootPrefixCls}--simple-text` : `${rootPrefixCls}--text`),
    onClick: function (e) {
      var _operationAreaRef$cur;
      if (!previewable) {
        return;
      }
      const {
        target
      } = e;
      // 防止操作按钮触发 item preview
      if ((target instanceof HTMLElement || target instanceof SVGElement) && (e.target === operationAreaRef.current || (_operationAreaRef$cur = operationAreaRef.current) !== null && _operationAreaRef$cur !== void 0 && _operationAreaRef$cur.contains(target))) {
        return;
      }
      onPreview === null || onPreview === void 0 ? void 0 : onPreview();
    },
    style: style,
    sx: [previewable && {
      "props": {
        "Bceei9c": u12o6wmt
      }
    }]
  }, jsx("div", {
    className: `${rootPrefixCls}-container`
  }, jsx("div", {
    className: `${rootPrefixCls}-icon`
  }, iconRender === null || iconRender === void 0 ? void 0 : iconRender(file)), jsx("div", {
    className: `${rootPrefixCls}-info`
  }, jsx(FileName, {
    file: file,
    nameRenderer: nameRenderer,
    className: `${rootPrefixCls}-info-name`
  }), !isSimple && jsx(FileStatus, {
    file: file,
    prefixCls: rootPrefixCls
  })), jsx("div", {
    className: `${rootPrefixCls}-operations`,
    ref: operationAreaRef
  }, isSimple && errorNode, action)), jsx(UploadProgress, {
    file: file,
    progressProps: progressProps,
    prefixCls: rootPrefixCls
  })));
}
