// @universe-design/react/es/upload/uploadFile/useFile.js
//.u114py8l{color:@function-danger-content-default;}
import ulvoz0o from "./../../.stylex/ulvoz0o";
import * as React from 'react';
import DeleteTrashOutlined from '@universe-design/icons-react/DeleteTrashOutlined';
import DownloadOutlined from '@universe-design/icons-react/DownloadOutlined';
import RefreshOutlined from '@universe-design/icons-react/RefreshOutlined';
import VisibleOutlined from '@universe-design/icons-react/VisibleOutlined';
import WarningFilled from '@universe-design/icons-react/WarningFilled';
import stylex from '@universe-design/stylex';
import clsx from 'clsx';
import { useCssScope, useLocaleReceiver } from '../../config-provider/hooks';
import { internalPreview, internalDownload, isError } from '../utils';
import { Tooltip } from '../../tooltip';
import { IconButton } from '../../button';
import getFileIcon from '../utils/getFileIcon';
import { useFileContext } from './FileContext';
export function OperationIconComp(_ref) {
  let {
    icon,
    handleClick,
    tooltip,
    className,
    style,
    color
  } = _ref;
  const {
    file,
    rootPrefixCls: prefixCls
  } = useFileContext();
  if (typeof icon === 'function') {
    icon = icon(file);
  }
  if (!icon) {
    return null;
  }
  return /*#__PURE__*/React.createElement(Tooltip, {
    title: tooltip
  }, /*#__PURE__*/React.createElement(IconButton, {
    className: clsx(`${prefixCls}-operation-icon`, className),
    icon: icon,
    onClick: () => handleClick === null || handleClick === void 0 ? void 0 : handleClick(file),
    style: style,
    size: "small",
    color: color
  }));
}
export function useFile(props) {
  const {
    file: fileProp,
    prefixCls,
    className
  } = props;
  const file = React.useMemo(() => ({
    status: 'success',
    ...fileProp
  }), [fileProp]);
  const defaultFileProps = React.useMemo(() => ({
    disabled: false,
    progress: {
      strokeWidth: 4,
      showInfo: false
    },
    showRemoveIcon: true,
    showDownloadIcon: true,
    showPreviewIcon: true,
    showErrorIcon: true,
    showRetryIcon: true,
    removeIcon: /*#__PURE__*/React.createElement(DeleteTrashOutlined, null),
    onPreview: () => internalPreview(file),
    previewIcon: /*#__PURE__*/React.createElement(VisibleOutlined, null),
    onDownload: () => internalDownload(file),
    downloadIcon: /*#__PURE__*/React.createElement(DownloadOutlined, null),
    retryIcon: /*#__PURE__*/React.createElement(RefreshOutlined, null),
    errorIcon: /*#__PURE__*/React.createElement(WarningFilled, {
      className: stylex({
        "props": {
          "sj55zd": ulvoz0o
        }
      })
    })
  }), [file]);
  const mergedFileProps = {
    // 默认值优先级最低
    ...defaultFileProps,
    // props 优先级最高
    ...props
  };
  const {
    showDownloadIcon,
    downloadIcon,
    showErrorIcon,
    showPreviewIcon,
    showRemoveIcon,
    showRetryIcon,
    errorIcon,
    previewIcon,
    progress,
    disabled,
    removeIcon,
    retryIcon,
    onRemove,
    onRetry,
    onDownload,
    onPreview,
    iconRender: iconRenderProp
  } = mergedFileProps;
  const rootPrefixCls = useCssScope('upload-item', prefixCls)();
  const locale = useLocaleReceiver('upload');
  const status = file.status || 'success';
  const downloadNode = React.useMemo(() => {
    if (!showDownloadIcon || status !== 'success') {
      return null;
    }
    return /*#__PURE__*/React.createElement(OperationIconComp, {
      icon: downloadIcon,
      handleClick: onDownload,
      tooltip: locale.download
    });
  }, [downloadIcon, locale.download, onDownload, showDownloadIcon, status]);
  const errorNode = React.useMemo(() => {
    if (!showErrorIcon || status !== 'error') {
      return null;
    }
    const failedReason = file.failedReason || locale.failed;
    return /*#__PURE__*/React.createElement(OperationIconComp, {
      icon: errorIcon,
      tooltip: failedReason
    });
  }, [errorIcon, file, locale.failed, showErrorIcon, status]);
  const previewable = showPreviewIcon && ['success', 'downloaded'].includes(status);
  const previewNode = React.useMemo(() => {
    if (!previewable || !previewIcon) {
      return null;
    }
    return /*#__PURE__*/React.createElement(OperationIconComp, {
      icon: previewIcon,
      tooltip: locale.preview,
      handleClick: onPreview
    });
  }, [locale.preview, onPreview, previewable, previewIcon]);
  const retryNode = React.useMemo(() => {
    if (!showRetryIcon || status !== 'error' || disabled || !retryIcon) {
      return null;
    }
    return /*#__PURE__*/React.createElement(OperationIconComp, {
      icon: retryIcon,
      tooltip: locale.uploadAgain,
      handleClick: onRetry,
      className: `${rootPrefixCls}-operation-retry`,
      color: "primary"
    });
  }, [disabled, locale.uploadAgain, onRetry, retryIcon, showRetryIcon, status, rootPrefixCls]);
  const removeNode = React.useMemo(() => {
    if (!showRemoveIcon || disabled || file.status === 'downloading') {
      return null;
    }
    return /*#__PURE__*/React.createElement(OperationIconComp, {
      icon: removeIcon,
      tooltip: locale.delete,
      handleClick: onRemove
    });
  }, [disabled, file, locale.delete, onRemove, removeIcon, showRemoveIcon]);
  const classNames = clsx(className, rootPrefixCls, `${rootPrefixCls}--${file.status}`, isError(file) ? `${rootPrefixCls}--error` : `${rootPrefixCls}--normal`);
  return {
    downloadNode,
    errorNode,
    removeNode,
    previewNode,
    retryNode,
    showPreviewIcon,
    showErrorIcon,
    iconRender: function (f) {
      var _iconRenderProp;
      return (_iconRenderProp = iconRenderProp === null || iconRenderProp === void 0 ? void 0 : iconRenderProp(f)) !== null && _iconRenderProp !== void 0 ? _iconRenderProp : getFileIcon(f);
    },
    onRemove,
    onPreview,
    file,
    disabled,
    rootPrefixCls,
    progressProps: progress,
    classNames
  };
}
