// @universe-design/react/src/form/FormItemUI.tsx
import InfoOutlined from '@universe-design/icons-react/InfoOutlined';
import clsx from 'clsx';
import type { CSSProperties, ReactNode } from 'react';
import React from 'react';
import stylex from '@universe-design/stylex';
import type { ColProps } from '../col';
import { Col } from '../col';
import type { FieldProps } from '../react-mini-form';
import { Row } from '../row';
import { useCssScope } from '../config-provider/hooks';
import type { TooltipProps } from '../tooltip';
import { Tooltip } from '../tooltip';
import { isFunction, isNonNullable } from '../util';
import { useFormUIContextSelector } from './FormUIContext';
import { forwardHTMLRef } from '../util/forwardRef';
import { Text } from '../text';

import './style';
/**
 * @public
 */
export type WrapperTooltipProps = TooltipProps & {
  icon?: React.ReactElement;
};
/**
 * @public
 */
export type LabelTooltipType = WrapperTooltipProps | React.ReactNode;

/**
 * @public
 * @docgen
 */
export interface FormItemUIProps extends Pick<FieldProps, 'validateStatus'> {
  children: ReactNode;
  /**
   * 额外的提示信息，和 `help` 类似，当需要错误信息和提示文案同时出现时，可以使用这个。
   */
  extra?: ReactNode;
  /**
   * 是否隐藏字段（依然会收集和校验字段）。
   * @defaultValue false
   */
  hidden?: boolean;
  /**
   * 设置子元素 Label `htmlFor` 属性。
   */
  htmlFor?: string;
  /**
   * Label 标签的文本。
   */
  label?: ReactNode;
  /**
   * 标签文本对齐方式。
   * @defaultValue `right`
   */
  labelAlign?: 'left' | 'right';
  /**
   * Label 标签布局，同 Col 组件，设置 `span offset` 值，如 `{span: 3, offset: 12}` 或 `sm: {span: 3, offset: 12}`。你可以通过 Form 的 `labelCol` 进行统一设置，不会作用于嵌套 Item。当和 `Form` 同时设置时，以 `Item` 为准。
   */
  labelCol?: ColProps;
  /**
   * 是否显示必选标识
   * @defaultValue false
   */
  required?: boolean;
  /**
   * 配置提示信息。
   */
  tooltip?: LabelTooltipType;
  /**
   * 需要为输入控件设置布局样式时，使用该属性，用法同 `labelCol`。你可以通过 Form 的 `wrapperCol` 进行统一设置，不会作用于嵌套 Item。当和 Form 同时设置时，以 Item 为准。
   */
  wrapperCol?: ColProps;
  className?: string;
  style?: CSSProperties;
  /**
   * 提示信息，如不设置，则会根据校验规则自动生成。
   */
  help?: ReactNode;
  /**
   * 自定义 Label 提示图标，主要用于自定义弹出组件
   */
  labelInfoIconRenderer?: (infoIconNode: React.ReactElement) => React.ReactNode;
  /**
   * 表单报错展示方式，共有三种。
   * - `locally` 代表「原位报错」，表单项初始间距较大，错误信息会在表单项间距处出现，不额外占位。
   * - `squeezed` 代表「挤压报错」，表单项初始间距较小，错误信息会额外占位。
   * - `none` 代表「报错不展示」，表单项只有报错状态，不展示报错信息
   * @defaultValue `locally`
   */
  errorDisplay?: 'locally' | 'squeezed' | 'none';
}

export const FormItemUI = forwardHTMLRef<FormItemUIProps, 'div'>((props, ref) => {
  const {
    label,
    className,
    style,
    children,
    labelAlign,
    labelCol,
    required = false,
    wrapperCol,
    help,
    hidden,
    validateStatus,
    tooltip,
    extra,
    htmlFor,
    labelInfoIconRenderer,
    errorDisplay: errorDisplayProp,
    ...restProps
  } = props;
  const formCtxLabelAlign = useFormUIContextSelector(ctx => ctx.labelAlign);
  const formCtxLabelCol = useFormUIContextSelector(ctx => ctx.labelCol);
  const formCtxWrapperCol = useFormUIContextSelector(ctx => ctx.wrapperCol);
  const formCtxErrorDisplay = useFormUIContextSelector(ctx => ctx.errorDisplay);
  const mergedLabelAlign = labelAlign ?? formCtxLabelAlign ?? 'right';
  const mergedLabelCol = labelCol ?? formCtxLabelCol ?? {};
  const mergedWrapperCol = wrapperCol ?? formCtxWrapperCol ?? {};
  const errorDisplay = errorDisplayProp ?? formCtxErrorDisplay ?? 'locally';
  const itemCssScope = useCssScope('form__item');
  const itemLabelCssScope = useCssScope('form__item__label');
  const itemControlCssScope = useCssScope('form__item__control');
  const hasExplain = isNonNullable(help) && errorDisplay !== 'none';
  const isError = validateStatus === 'error';

  /**
   * ================ Tooltip Content Node ================
   */
  const { tooltipProps, labelInfoIconNode } = ((): {
    tooltipProps: TooltipProps;
    labelInfoIconNode: React.ReactElement;
  } => {
    let _tooltipProps = {} as WrapperTooltipProps;
    if (tooltip && typeof tooltip === 'object' && !React.isValidElement(tooltip)) {
      _tooltipProps = tooltip as WrapperTooltipProps;
    } else {
      _tooltipProps = { title: tooltip };
    }
    const { icon = <InfoOutlined />, ...restTooltipProps } = _tooltipProps;
    const _labelInfoIconNode = (
      <Text as="div" className={itemCssScope('__tooltip')}>
        {icon}
      </Text>
    );
    return {
      tooltipProps: restTooltipProps,
      labelInfoIconNode: _labelInfoIconNode,
    };
  })();

  const tooltipNode = (() => {
    if (isFunction(labelInfoIconRenderer)) {
      return labelInfoIconRenderer(labelInfoIconNode);
    }
    if (tooltip) {
      return <Tooltip {...tooltipProps}>{labelInfoIconNode}</Tooltip>;
    }
    return null;
  })();

  /**
   * ================ Extra Node ================
   */
  const extraNode = extra ? <Text className={itemControlCssScope('__extra')}>{extra}</Text> : null;

  /**
   * ================ Error Or Help Node ================
   */
  const errorOrHelpNode = hasExplain ? (
    <Text className={clsx(itemControlCssScope('__explain'), isError && itemControlCssScope('__explain-error'))}>
      {help}
    </Text>
  ) : null;

  return (
    <Row
      {...restProps}
      className={clsx(
        itemCssScope(),
        hasExplain && itemCssScope('-with-help'),
        hidden && itemCssScope('-hidden'),
        itemCssScope(`-error-${errorDisplay}`),
        className,
      )}
      style={style}
      ref={ref}
    >
      {isNonNullable(label) ? (
        <Col
          {...mergedLabelCol}
          className={clsx(
            itemLabelCssScope(),
            mergedLabelAlign === 'left' && itemLabelCssScope('-left'),
            mergedLabelCol.className,
          )}
        >
          <Text
            as="label"
            htmlFor={htmlFor}
            title={typeof label === 'string' ? label : undefined}
            lineClamp={2}
            sx={[stylex.create({ minWidth: 'min-content' })]}
          >
            {label}
          </Text>
          {required && <Text className={itemCssScope('__required-mark')}>*</Text>}
          {tooltipNode}
        </Col>
      ) : null}
      <Col {...mergedWrapperCol} className={clsx(itemControlCssScope(), mergedWrapperCol.className)}>
        <div className={itemControlCssScope('__input')}>
          <div className={itemControlCssScope('__input__content')}>{children}</div>
        </div>
        {errorOrHelpNode}
        {extraNode}
      </Col>
    </Row>
  );
});

if (__DEV__) {
  FormItemUI.displayName = 'FormItemUI';
}
