// @universe-design/react/src/form/index.tsx
/* eslint-disable react/destructuring-assignment */
import clsx from 'clsx';
import type { ReactNode, ReactElement, RefAttributes } from 'react';
import React, { isValidElement, forwardRef, memo, useMemo, useImperativeHandle } from 'react';
import type { Optional } from 'utility-types';
import merge from 'lodash/merge';
import get from 'lodash/get';
import type { Options as ScrollIntoViewIfNeededOptions } from 'scroll-into-view-if-needed';
import scrollIntoView from 'scroll-into-view-if-needed';
import type {
  FormProps as MiniFormProps,
  FieldProps,
  FormContextType as MiniFormContextType,
  UseFieldReturnValue,
  Validator,
  FormListProps as MiniFormListProps,
  FormContextType,
} from '../react-mini-form';
import {
  Field,
  useFieldContext,
  marshalNamePath,
  FieldPropsAreEqual,
  FormList as MiniFormList,
  useForm,
  FormContextProvider,
} from '../react-mini-form';
import { useComponentSize, useCssScope, useLocaleReceiver } from '../config-provider/hooks';
import type { TooltipProps } from '../tooltip';
import { devWarning, isFunction, isNonNullable, getDefaultSizeAbbr, warning } from '../util';
import type { RuleItem } from 'async-validator';
import { validateRule } from './utils/validateUtil';
import type { FormUIProps } from './FormUIContext';
import { FormUIContextProvider, useFormUIContextSelector } from './FormUIContext';
import type { FormItemUIProps } from './FormItemUI';
import { FormItemUI } from './FormItemUI';
import type { ObjectValidator, ValidateTrigger } from '../react-mini-form/shared';
import { unmarshalNamePath, asArray } from '../react-mini-form/shared';
import { ConfigProvider } from '../config-provider';
import { FindDOMNode } from '@universe-design/react-utilities';
import { FormDescendantsProvider, useFormDescendant, useFormDescendants } from './FormDescendantContext';

import './style';
export type { ValidateMessages } from './interface';
/**
 * @public
 */
export type WrapperTooltipProps = TooltipProps & {
  icon?: React.ReactElement;
};

/**
 * @public
 * @docgen
 */
export interface EnhancedRuleItem extends RuleItem {
  /**
   * 校验触发时机
   */
  validateTrigger?: ValidateTrigger | ValidateTrigger[];
}
export interface FormItemControlProps extends FormItemUIProps, Omit<FieldProps, 'children'> {
  /**
   * Field 子节点, 可以是 `ReactNode` 或是一个 `RenderProp`。
   */
  children: ReactNode | ((props: UseFieldReturnValue) => ReactNode);
  /**
   * 是否隐藏字段（依然会收集和校验字段）。
   * @defaultValue false
   */
  hidden?: boolean;
  /**
   * 为 `true` 时不带样式，作为纯字段控件使用。
   * @defaultValue false
   */
  noStyle?: boolean;
  /**
   * 必填样式设置。如不设置，则会根据校验规则自动生成。
   * @defaultValue false
   */
  required?: boolean;
  /**
   * 校验规则，设置字段的校验逻辑。
   */
  rules?: EnhancedRuleItem | EnhancedRuleItem[];
}
/**
 * @docgen
 * @public
 * 表单字段组件`Form.Item`，用于数据双向绑定、校验、布局等。
 */
export type FormItemProps = Optional<FormItemControlProps, 'name'>;

/**
 * @docgen
 * @public
 * @typeParam Values - Values of this form
 */
export interface FormProps<Values> extends Omit<MiniFormProps<Values>, 'onSubmit'>, FormUIProps {
  /**
   * 通过 Validation 之后的提交过程回调
   */
  onFinish?: MiniFormProps<Values>['onSubmit'];
  /**
   * 是否在提交校验失败后滚到第一个错误表单项
   * @defaultValue true
   */
  shouldScrollToFirstErrorField?: ScrollIntoViewIfNeededOptions | boolean;
}

/**
 * @public
 */
export interface FormRef<Values> extends FormContextType<Values> {
  /**
   * 将第一个报错项滚动到可视区域
   */
  scrollToFirstError: (option?: ScrollIntoViewIfNeededOptions) => void;
}

/**
 * @docgen
 * @public
 * @typeParam FieldValue - FieldValue of this form list
 */
export type FormListProps<FieldValue> = MiniFormListProps<FieldValue> & Pick<FormItemProps, 'rules'>;

const BaseForm = forwardRef<FormRef<any>, FormProps<any>>(function Form(props, ref) {
  const size = useComponentSize();
  const {
    layout = 'vertical',
    className,
    requiredMark = true,
    size: sizeProp = size,
    labelAlign,
    labelCol,
    wrapperCol,
    onFinish,
    validateMessages,
    errorDisplay,
    component = 'form',
    children,
    shouldScrollToFirstErrorField = true,
    onSubmitError,
    ...restProps
  } = props;

  /**
   * ================ ClassName ================
   */
  const formCssScope = useCssScope('form');

  const formClassName = clsx(
    formCssScope(),
    formCssScope(`-${layout}`),
    !requiredMark && formCssScope('-hide-required-mark'),
    formCssScope(`-${getDefaultSizeAbbr(sizeProp, 'medium')}`),
    className,
  );

  /**
   * ================ Logic context ================
   */
  const formCtx = useForm({
    ...restProps,
    onSubmit: onFinish,
    onSubmitError: errors => {
      isFunction(onSubmitError) && onSubmitError(errors);
      // eslint-disable-next-line @typescript-eslint/no-use-before-define
      scrollToFirstError();
    },
  });
  const { submit } = formCtx;

  /**
   * ================ Descendants ================
   */
  const descendants = useFormDescendants();

  /**
   * ================ Methods ================
   */

  // ---------------- Handle submit ----------------
  const handleSubmit = React.useCallback<React.FormEventHandler<HTMLFormElement>>(
    e => {
      e.preventDefault();
      e.stopPropagation();

      submit();
    },
    [submit],
  );
  let childNode = children;
  if (component) {
    childNode = <form {...{ onSubmit: handleSubmit, className: formClassName }}>{childNode}</form>;
  }

  // ---------------- scrollToFirstError ----------------
  const scrollToFirstError = React.useCallback(
    (option?: ScrollIntoViewIfNeededOptions) => {
      if (shouldScrollToFirstErrorField && formCtx.isDirty) {
        const node = descendants.values().find(o => o.id && get(formCtx.errors, unmarshalNamePath(o.id)))?.node;
        if (node) {
          const mergedOption =
            option ??
            (shouldScrollToFirstErrorField === true ? { block: 'nearest' as const } : shouldScrollToFirstErrorField);

          scrollIntoView(node, {
            scrollMode: 'if-needed',
            ...(mergedOption || {}),
          });
        }
      }
    },
    [descendants, formCtx.errors, formCtx.isDirty, shouldScrollToFirstErrorField],
  );

  /**
   * ================ Ref ================
   */
  useImperativeHandle(ref, () =>
    Object.assign(Object.create(Object.getPrototypeOf(formCtx), Object.getOwnPropertyDescriptors(formCtx)), {
      scrollToFirstError,
    }),
  );

  return (
    <FormDescendantsProvider value={descendants}>
      <ConfigProvider componentSize={sizeProp}>
        <FormUIContextProvider
          value={{
            requiredMark,
            labelAlign,
            labelCol,
            wrapperCol,
            validateMessages,
            errorDisplay,
          }}
        >
          <FormContextProvider value={formCtx}>{childNode}</FormContextProvider>
        </FormUIContextProvider>
      </ConfigProvider>
    </FormDescendantsProvider>
  );
});

type UseValidatorsProps = Pick<FormItemProps, 'label' | 'rules' | 'validate'>;

function useValidators(props: UseValidatorsProps) {
  const { label, rules, validate: validateProp } = props;
  const formCtxValidateMessages = useFormUIContextSelector(ctx => ctx.validateMessages);
  const locale = useLocaleReceiver('form');
  const variables = useMemo(() => {
    const _variables: Record<string, string> = {};
    if (typeof label === 'string') {
      _variables.label = label;
    } else {
      _variables.label = '';
    }
    return _variables;
  }, [label]);

  const validate = React.useMemo<Validator[]>(() => {
    const validators: Validator[] = [];
    if (rules) {
      validators.push(
        ...asArray(rules).map(rule => {
          const validator: ObjectValidator = {
            validateTrigger: rule.validateTrigger,
            validate: async (value, _name) => {
              const namePathString = marshalNamePath(_name);
              const errorMessages = await validateRule(
                namePathString,
                value,
                rule,
                {
                  validateMessages: merge({}, formCtxValidateMessages, locale.defaultValidateMessage),
                },
                variables,
              );
              if (errorMessages.length > 0) {
                return errorMessages[0];
              }
            },
          };
          return validator;
        }),
      );
    }
    if (validateProp) {
      validators.push(...asArray(validateProp));
    }
    return validators;
  }, [locale.defaultValidateMessage, formCtxValidateMessages, rules, validateProp, variables]);

  return validate;
}

function _FormItem(props: FormItemProps) {
  const {
    noStyle = false,
    label,
    className,
    style,
    name,
    children,
    labelAlign,
    labelCol,
    rules,
    // eslint-disable-next-line no-nested-ternary
    required = rules ? (Array.isArray(rules) ? rules.some(r => r.required) : rules.required) : undefined,
    wrapperCol,
    help,
    validateStatus,
    tooltip,
    extra,
    hidden,
    htmlFor,
    validate: validateProp,
    labelInfoIconRenderer,
    errorDisplay,
    ...restProps
  } = props;
  const { error } = useFieldContext({ name });

  const mergedValidateStatus = validateStatus ?? (error ? 'error' : undefined);

  warning(
    isNonNullable(name),
    devWarning('FormItem', 'The `name` prop is undefined, thus this field will not be registered to the form'),
  );

  const validate = useValidators({ label, validate: validateProp, rules });

  /**
   * Descendant
   */
  const id = isNonNullable(name) ? marshalNamePath(name) : undefined;
  const { register } = useFormDescendant(
    React.useMemo(
      () => ({
        id,
      }),
      [id],
    ),
  );

  /**
   * ================ Input Content Node ================
   *
   * child type: funcChild | reactNodeChild
   *
   * noStyle: true | false
   *
   * - 如果 noStyle === false，则直接用 FormItemUI 的 ref 来 register
   * - 如果 noStyle === true，则需要判断 children 类型
   *  - 如果 children 是 ReactElement 则直接套住 FindDOMNode
   *  - 如果 children 是 function 则先交给 Field 处理，然后再套住 FindDOMNode
   */
  let inputContentNode: ReactNode = children;
  // 判断是否是一个 field 组件，如果没有 name 的话则认为只是一个 UI submit 而不是 field
  if (props.name !== undefined && (isValidElement(children) || isFunction(children))) {
    inputContentNode = (
      <Field name={props.name} validate={validate} validateStatus={mergedValidateStatus} {...restProps}>
        {(() => {
          if (!noStyle) {
            return children;
          }
          if (isFunction(children)) {
            return fieldProps => (
              <FindDOMNode domOnly childRef={register}>
                {children(fieldProps)}
              </FindDOMNode>
            );
          }
          return (
            <FindDOMNode {...(children.props as Record<string, unknown>)} domOnly childRef={register}>
              {children}
            </FindDOMNode>
          );
        })()}
      </Field>
    );
  }

  if (noStyle) {
    return <>{inputContentNode}</>;
  }

  return (
    <FormItemUI
      {...{
        label,
        className,
        style,
        children,
        labelAlign,
        labelCol,
        required,
        wrapperCol,
        help: help ?? error,
        hidden,
        validateStatus: mergedValidateStatus,
        tooltip,
        extra,
        htmlFor,
        labelInfoIconRenderer,
        errorDisplay,
        ref: register,
      }}
    >
      {inputContentNode}
    </FormItemUI>
  );
}

export function FormList<FieldValue>(props: FormListProps<FieldValue>) {
  const { validate, rules, ...others } = props;
  const validators = useValidators({
    rules,
    validate,
  });

  return <MiniFormList validate={validators} {...others} />;
}

export type {
  FormContextType,
  Validator,
  FunctionValidator,
  ObjectValidator,
  ContextSelector,
} from '../react-mini-form';
export {
  useField,
  useFieldContext,
  useForm,
  useFormContext,
  useFormContextSelector,
  FormSelector,
  FormSelectors,
  useFormList,
  FormContextProvider,
} from '../react-mini-form';
export type {
  UseFormProps,
  FormErrors,
  FormTouched,
  UseFormListProps,
  Path,
  MultiLevelPath,
  OneLevelPath,
  UseFieldReturnValue,
  UseFieldProps,
  UseFormListReturnValue,
  FieldProps,
  FormListField,
  FormListActions,
  FormInternalContextType,
} from '../react-mini-form';
export type { MiniFormListProps };

export const Form = BaseForm as unknown as (<Values>(
  props: FormProps<Values> & React.RefAttributes<MiniFormContextType<Values>>,
) => ReactElement<any, any> | null) & {
  Item: typeof _FormItem;
  ItemUI: typeof FormItemUI;
};

export const FormItem = memo(_FormItem, FieldPropsAreEqual);
Form.Item = FormItem as typeof _FormItem;
Form.ItemUI = FormItemUI;

if (__DEV__) {
  FormItem.displayName = 'FormItem';
}

export type { LabelTooltipType, FormItemUIProps } from './FormItemUI';
export type { FormUIProps } from './FormUIContext';
