// @universe-design/react/src/form/interface.ts
import type { IntlMessage } from '../shared/locales/types';

/**
 *  参考 [code](https://code.byted.org/universe-ui/react/tree/master/@universe-design/react/src/form/interface.ts)
 * @privateRemarks
 * fork from async-validator
 */
export interface ValidateMessages {
  default: IntlMessage;
  required: IntlMessage;
  enum: IntlMessage;
  whitespace: IntlMessage;
  date: {
    format: IntlMessage;
    parse: IntlMessage;
    invalid: IntlMessage;
  };
  types: {
    string: IntlMessage;
    method: IntlMessage;
    array: IntlMessage;
    object: IntlMessage;
    number: IntlMessage;
    date: IntlMessage;
    boolean: IntlMessage;
    integer: IntlMessage;
    float: IntlMessage;
    regexp: IntlMessage;
    email: IntlMessage;
    url: IntlMessage;
    hex: IntlMessage;
  };
  string: {
    len: IntlMessage;
    min: IntlMessage;
    max: IntlMessage;
    range: IntlMessage;
  };
  number: {
    len: IntlMessage;
    min: IntlMessage;
    max: IntlMessage;
    range: IntlMessage;
  };
  array: {
    len: IntlMessage;
    min: IntlMessage;
    max: IntlMessage;
    range: IntlMessage;
  };
  pattern: {
    mismatch: IntlMessage;
  };
}
