// @universe-design/react/src/layout/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import CBasic from '../demo/basic';
import CResponsive from '../demo/responsive';
import CSider from '../demo/sider';
import CTopSider_2 from '../demo/top-sider-2';
import CTopSider from '../demo/top-sider';
import CTop from '../demo/top';
if (!(CBasic as Story).storyName) {
  (CBasic as Story).storyName = '0. 基础用法';
  (CBasic as Story).parameters = {
    docs: {
      description: {
        story:
          '经典的页面布局。\n\n\n<style>\n  .ud__layout-demo-basic {\n    text-align: center;\n  }\n\n  .ud__layout-demo-basic .ud__layout__header,\n  .ud__layout-demo-basic .ud__layout__footer {\n    background: var(--B300);\n    color: #fff;\n  }\n\n  .ud__layout-demo-basic .ud__layout__sider {\n    background: var(--B400);\n    color: #fff;\n    line-height: 120px;\n  }\n\n  .ud__layout-demo-basic .ud__layout__content {\n    background: var(--B500);\n    color: #fff;\n    min-height: 100px;\n    line-height: 120px;\n  }\n\n  .ud__layout-demo-basic > .ud__layout + .ud__layout {\n    margin-top: 48px;\n  }\n</style>\n',
      },
    },
  };
}
if (!(CResponsive as Story).storyName) {
  (CResponsive as Story).storyName = '5. 响应式侧边栏';
  (CResponsive as Story).parameters = {
    docs: {
      description: {
        story:
          '设置`breakpoint`可触发响应式收缩。\n\n\n<style>\n\n  .ud__layout-demo-responsive {\n    border: 1px solid var(--line-border-card);\n  }\n\n  .ud__layout-demo-responsive .logo {\n    height: 32px;\n    background: var(--bg-mask);\n    margin: 16px;\n  }\n\n  .site-layout-sub-header-background {\n    background: var(--bg-float);\n  }\n\n  .site-layout-background {\n    background: var(--bg-float);\n  }\n</style>\n',
      },
    },
  };
}
if (!(CSider as Story).storyName) {
  (CSider as Story).storyName = '4. 侧边布局';
  (CSider as Story).parameters = {
    docs: {
      description: {
        story: '侧边两列式布局。页面横向空间有限时，侧边导航可收起。\n\n',
      },
    },
  };
}
if (!(CTopSider_2 as Story).storyName) {
  (CTopSider_2 as Story).storyName = '3. 顶部 - 侧边布局 - 通栏';
  (CTopSider_2 as Story).parameters = {
    docs: {
      description: {
        story: '同样拥有顶部导航及侧边栏，区别是两边未留边距，多用于应用型的网站。\n\n',
      },
    },
  };
}
if (!(CTopSider as Story).storyName) {
  (CTopSider as Story).storyName = '2. 顶部 - 侧边布局';
  (CTopSider as Story).parameters = {
    docs: {
      description: {
        story: '拥有顶部导航及侧边栏的页面，多用于展示类网站。\n\n',
      },
    },
  };
}
if (!(CTop as Story).storyName) {
  (CTop as Story).storyName = '1. 上中下布局';
  (CTop as Story).parameters = {
    docs: {
      description: {
        story: '最基本的「上-中-下」布局。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Layout/官网示例',
  tags: ['autodocs'],
};

export { CBasic, CResponsive, CSider, CTopSider_2, CTopSider, CTop };
