// @universe-design/react/src/collapse/stories/screenshot.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import CAccordion from '../demo/accordion';
import CBasic from '../demo/basic';
import CCollapsible from '../demo/collapsible';
import CCustom from '../demo/custom';
import CExtra from '../demo/extra';
import CIg from '../demo/ig';
import CNest from '../demo/nest';
if (!(CAccordion as Story).storyName) {
  (CAccordion as Story).storyName = 'Accordion';
}
if (!(CBasic as Story).storyName) {
  (CBasic as Story).storyName = 'Basic';
}
if (!(CCollapsible as Story).storyName) {
  (CCollapsible as Story).storyName = 'Collapsible';
}
if (!(CCustom as Story).storyName) {
  (CCustom as Story).storyName = 'Custom';
}
if (!(CExtra as Story).storyName) {
  (CExtra as Story).storyName = 'Extra';
}
if (!(CIg as Story).storyName) {
  (CIg as Story).storyName = 'Ig';
}
if (!(CNest as Story).storyName) {
  (CNest as Story).storyName = 'Nest';
}

export default {
  title: 'Components/Collapse/截图测试',
  tags: ['autodocs'],
};

export { CAccordion, CBasic, CCollapsible, CCustom, CExtra, CIg, CNest };
