// @byted-service/consul/lib/fns.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.random = exports.lookupName = exports.lookup = void 0;
const env_1 = require("@byted-service/env");
const consul_1 = require("./consul");
const errors_1 = require("./errors");
const defaultConsul = new consul_1.Consul();
const ipv4 = process.env.BYTED_HOST_IP;
const ipv6 = process.env.BYTED_HOST_IPV6;
const lookup = async (psm, options = {}, consul = defaultConsul) => {
    const endpoints = await consul.lookup(psm, options);
    if (endpoints.length === 0) {
        throw new errors_1.ConsulError('No endpoint. PSM: %s, options: %j, address: %s, environments: %j', psm, options, consul.address, {
            CONSUL_HTTP_HOST: process.env.CONSUL_HTTP_HOST || '',
            MY_HOST_IP: process.env.MY_HOST_IP || '',
            TCE_HOST_IP: process.env.TCE_HOST_IP || '',
            MY_HOST_IPV6: process.env.MY_HOST_IPV6 || '',
            ROUTE_IP: process.env.ROUTE_IP || '',
        });
    }
    return endpoints;
};
exports.lookup = lookup;
const lookupName = async (psm, options = {}, consul = defaultConsul) => {
    if (ipv4 && ipv6) {
        options.addrfam = options.addrfam || 'dual-stack';
        options.unique = options.unique || 'v6';
    }
    else if (ipv6) {
        options.addrfam = options.addrfam || 'v6';
    }
    else if (ipv4) {
        options.addrfam = options.addrfam || 'v4';
    }
    if (!(0, env_1.isInTCE)()) {
        options.addrfam = options.addrfam || 'dual-stack';
        options.unique = options.unique || 'v6';
    }
    return await (0, exports.lookup)(psm, options, consul);
};
exports.lookupName = lookupName;
const random = async (psm, options = {}, consul = defaultConsul) => {
    const endpoints = await (0, exports.lookupName)(psm, options, consul);
    let total = 0;
    for (let i = 0; i < endpoints.length; i++) {
        total += endpoints[i].weight;
    }
    let rand = Math.floor(total * Math.random());
    for (let i = 0; i < endpoints.length; i++) {
        rand -= endpoints[i].weight;
        if (rand < 0) {
            return endpoints[i];
        }
    }
    /* istanbul ignore next */
    throw new errors_1.ConsulError('should never happen');
};
exports.random = random;
