// @universe-design/react/lib/date-picker/hooks/useFormatAndParser.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _isFunction = _interopRequireDefault(require("lodash/isFunction"));
var _hooks = require("../../config-provider/hooks");
var _moment = _interopRequireDefault(require("moment"));
/**
 * 将 input 的输入值转换为 date
 * 将 value 转换为 string 提供给 input
 */

function getDefaultFormat(options) {
  const {
    mode,
    showTime = false,
    presetFormats
  } = options;
  switch (mode) {
    case 'year':
      return presetFormats.year;
    case 'month':
      return presetFormats.month;
    case 'date':
    default:
      return showTime ? `${presetFormats.date} HH:mm:ss` : presetFormats.date;
  }
}
var _default = _ref => {
  let {
    format: formatProp,
    parse,
    mode,
    showTime
  } = _ref;
  const {
    locale: localeName,
    components: {
      datePicker: locale
    }
  } = (0, _hooks.useLocale)();
  const presetFormats = {
    year: 'YYYY',
    month: locale.longMonthFormat,
    date: locale.longDateFormat
  };
  const defaultFormat = getDefaultFormat({
    mode,
    showTime,
    presetFormats
  });
  const format = formatProp !== null && formatProp !== void 0 ? formatProp : defaultFormat;
  return {
    formatValue: function (value) {
      if (!value) {
        return '';
      }
      if ((0, _isFunction.default)(format)) {
        return format(value);
      }
      return value.clone().locale(localeName).format(format);
    },
    parseValue: function (value) {
      if (!value) {
        return null;
      }
      if ((0, _isFunction.default)(parse)) {
        return parse(value);
      }
      /**
       * 对于 parser 而言，如果 format 类型为 (value: moment) => string
       * 此时是无法处理的 （设计上的一个漏洞）
       * 此时回退到 defaultFormat
       */
      const formatForParser = typeof format === 'string' ? format : defaultFormat;
      return (0, _moment.default)(value, formatForParser, localeName, true);
    },
    format
  };
};
exports.default = _default;
