// @universe-design/react/src/input-number/stories/homeless.stories.tsx
import React, { useState } from 'react';
import { InputNumber } from '..';
import { Space, Switch, Divider, Tooltip, Button, Input, Form, Slider } from '../..';
import './index.less';

export default {
  title: 'Components/InputNumber/临时示例',
  component: InputNumber,
};

function onChange(value?: number) {
  console.log('changed', value, typeof value);
}

function onPressEnter(event: React.KeyboardEvent<HTMLInputElement>) {
  console.log('onPressEnter', event);
}

export const Controlled = () => {
  const [value, setValue] = useState<number>();
  const onChangeFn = (v: any) => {
    console.log('onChange', v);
    setValue(v);
  };
  return (
    <InputNumber
      step={0.2}
      precision={1}
      max={10.1}
      min={0}
      placeholder="请输入"
      value={value}
      // formatter={v => `${v}%`}
      // parser={v => (v ? v.replace('%', '') : '')}
      onChange={onChangeFn}
    />
  );
};

export const DifferentStates = () => {
  const [disabled, setDisabled] = useState<boolean>(false);
  const [error, setError] = useState<boolean>(false);
  return (
    <Space direction="vertical">
      <div>
        <span>禁用：</span>
        <Switch checked={disabled} onChange={setDisabled} />
      </div>
      <div>
        <span>错误：</span>
        <Switch checked={error} onChange={setError} />
      </div>
      <div>
        <InputNumber
          step={0.2}
          precision={1}
          max={10.1}
          min={0}
          disabled={disabled}
          error={error}
          placeholder="请输入"
          onChange={onChange}
        />
        <br /> <br />
        <InputNumber
          step={0.2}
          precision={1}
          max={10.1}
          min={-10}
          disabled={disabled}
          error={error}
          placeholder="请输入"
          onChange={onChange}
        />
        <br /> <br />
        <InputNumber
          defaultValue={1}
          step={0.2}
          precision={1}
          max={10.1}
          min={0}
          disabled={disabled}
          error={error}
          onChange={onChange}
        />
        <br /> <br />
        <InputNumber
          disabled
          defaultValue={1}
          step={0.2}
          precision={1}
          max={10.1}
          min={0}
          error={error}
          onChange={onChange}
        />
        <br /> <br />
        <InputNumber
          readOnly
          defaultValue={1}
          step={0.2}
          precision={1}
          max={10.1}
          min={0}
          disabled={disabled}
          error={error}
          onChange={onChange}
        />
        <br /> <br />
        <InputNumber
          defaultValue={0}
          step={0.2}
          precision={1}
          max={0}
          min={-10}
          disabled={disabled}
          error={error}
          onChange={onChange}
        />
        <Divider orientation="left">Size</Divider>
        <InputNumber
          size="small"
          autoFocus
          defaultValue={1}
          step={0.2}
          precision={1}
          max={10.1}
          min={0}
          disabled={disabled}
          error={error}
        />
        <InputNumber
          size="medium"
          defaultValue={1}
          step={0.2}
          precision={1}
          max={10.1}
          min={0}
          disabled={disabled}
          error={error}
        />
        <InputNumber
          size="large"
          defaultValue={1}
          step={0.2}
          precision={1}
          max={10.1}
          min={0}
          disabled={disabled}
          error={error}
        />
        <Divider orientation="left">format</Divider>
        <InputNumber
          defaultValue={1000}
          formatter={value => `$ ${value}`.replace(/\B(?=(\d{3})+(?!\d))/g, ',')}
          parser={value => (value ? value.replace(/\$\s?|(,*)/g, '') : '')}
          onChange={onChange}
          onPressEnter={onPressEnter}
          disabled={disabled}
          error={error}
        />
        <br /> <br />
        <InputNumber
          defaultValue={100}
          min={0}
          max={100}
          formatter={value => `${value}%`}
          parser={value => (value ? value.replace('%', '') : '')}
          onChange={onChange}
          disabled={disabled}
          error={error}
        />
      </div>
    </Space>
  );
};

export const StepperStates = () => {
  const [disabled, setDisabled] = useState<boolean>(false);
  const [error, setError] = useState<boolean>(false);
  return (
    <Space direction="vertical">
      <div>
        <span>禁用：</span>
        <Switch checked={disabled} onChange={setDisabled} />
      </div>
      <div>
        <span>错误：</span>
        <Switch checked={error} onChange={setError} />
      </div>
      <div>
        <InputNumber
          mode="stepper"
          step={0.2}
          precision={1}
          max={10.1}
          min={0}
          placeholder="请输入"
          onChange={onChange}
          disabled={disabled}
          error={error}
        />
        <br /> <br />
        <InputNumber
          mode="stepper"
          step={0.2}
          precision={1}
          max={0}
          min={-10}
          defaultValue={0}
          placeholder="请输入"
          onChange={onChange}
          disabled={disabled}
          error={error}
        />
        <br /> <br />
        <InputNumber
          mode="stepper"
          defaultValue={1}
          step={0.2}
          precision={1}
          max={10.1}
          min={0}
          onChange={onChange}
          disabled={disabled}
          error={error}
        />
        <br /> <br />
        <InputNumber
          mode="stepper"
          disabled
          defaultValue={1}
          step={0.2}
          precision={1}
          max={10.1}
          min={0}
          onChange={onChange}
          error={error}
        />
        <br /> <br />
        <InputNumber
          mode="stepper"
          readOnly
          defaultValue={1}
          step={0.2}
          precision={1}
          max={10.1}
          min={0}
          onChange={onChange}
          disabled={disabled}
          error={error}
        />
        <br /> <br />
        <InputNumber
          mode="stepper"
          autoFocus
          defaultValue={1}
          step={0.2}
          precision={1}
          max={10.1}
          min={0}
          onChange={onChange}
          disabled={disabled}
          error={error}
        />
        <Divider orientation="left">Size</Divider>
        <InputNumber
          mode="stepper"
          size="small"
          defaultValue={1}
          step={0.2}
          precision={1}
          max={10.1}
          min={0}
          disabled={disabled}
          error={error}
        />
        <InputNumber
          mode="stepper"
          size="medium"
          defaultValue={1}
          step={0.2}
          precision={1}
          max={10.1}
          min={0}
          disabled={disabled}
          error={error}
        />
        <InputNumber
          mode="stepper"
          size="large"
          defaultValue={1}
          step={0.2}
          precision={1}
          max={10.1}
          min={0}
          disabled={disabled}
          error={error}
        />
        <Divider orientation="left">format</Divider>
        <InputNumber
          mode="stepper"
          defaultValue={1000}
          formatter={value => `$ ${value}`.replace(/\B(?=(\d{3})+(?!\d))/g, ',')}
          parser={value => (value ? value.replace(/\$\s?|(,*)/g, '') : '')}
          onChange={onChange}
          onPressEnter={onPressEnter}
          disabled={disabled}
          error={error}
        />
        <br /> <br />
        <InputNumber
          mode="stepper"
          defaultValue={100}
          min={0}
          max={100}
          formatter={value => `${value}%`}
          parser={value => (value ? value.replace('%', '') : '')}
          onChange={onChange}
          disabled={disabled}
          error={error}
        />
      </div>
    </Space>
  );
};

export const WidthTooltip = () => (
  <>
    <Tooltip title="hello" placement="bottom">
      <InputNumber
        step={0.2}
        precision={1}
        max={10.1}
        min={0}
        placeholder="请输入"
        onChange={onChange}
        onBlur={() => console.log('onBlur')}
        onFocus={() => console.log('onFocus')}
        onMouseEnter={() => console.log('onMouseEnter')}
        onMouseLeave={() => console.log('onMouseLeave')}
      />
    </Tooltip>
  </>
);

export const Demo = () => {
  const onChangeFn = (value?: number) => console.log('onChange', value);
  return (
    <InputNumber
      min={0}
      // max={10}
      precision={1}
      // step={0.1}
      // defaultValue={1}
      // value={2}
      onChange={onChangeFn}
    />
  );
};

export const Precision = () => (
  <InputNumber
    precision={3}
    defaultValue={0.0035}
    onChange={v => console.log('onChange', v)}
    onStep={(v, info) => console.log('onStep', v, info)}
  />
);

export const Error = () => {
  const [error1, setError1] = React.useState(false);
  const internalOnChange = (value: any) => console.log('changed', value);
  const [val1, setVal1] = React.useState<any>(1);
  const oRef1 = React.useRef<HTMLInputElement>(null);
  return (
    <>
      <Space size="large" direction="vertical">
        <div>
          <span>受控的 error: &nbsp;&nbsp;</span>
          <InputNumber error={error1} placeholder="请输入" onChange={internalOnChange} />
          &nbsp;&nbsp;
          <Button onClick={() => setError1(!error1)}>点我切换 error</Button>
        </div>
        --------------------------------------------------------
        <div>
          <span>自动把输入的字符过滤, 只保留数字: &nbsp;&nbsp;</span>
          <InputNumber onlyNumber value={val1} placeholder="请输入" onChange={(e: any) => setVal1(e)} />
          <br />
          <br />
          <Input ref={oRef1} placeholder="输入数字或字符, 然后点击下头按钮" style={{ width: 300 }} />
          <br />
          <br />
          <Button onClick={() => setVal1(oRef1.current?.value)}>先输入, 再点击, 然后改变</Button>
        </div>
        --------------------------------------------------------
        <div>
          <span>键入字符时自动报错: &nbsp;&nbsp;</span>
          <InputNumber min={-10} max={10} placeholder="请输入" onChange={internalOnChange} />
        </div>
        --------------------------------------------------------
        <div>
          <span>支持科学记数法: &nbsp;&nbsp;</span>
          <InputNumber placeholder="请输入" onChange={internalOnChange} />
        </div>
      </Space>
    </>
  );
};

export const Inform = () => (
  <Form layout="horizontal" initialValues={{ username: '', password: '', remember: true }}>
    <Form.Item
      required
      label="Username"
      name="username"
      rules={[{ required: true, message: 'Please input your username!' }]}
    >
      <InputNumber
        formatter={value => `$ ${value}`.replace(/\B(?=(\d{3})+(?!\d))/g, ',')}
        parser={value => value?.replace(/\$\s?|(,*)/g, '') || ''}
        precision={2}
      />
    </Form.Item>
  </Form>
);

export const LongWidth = () => {
  const [width, setWidth] = useState(0);
  const [value, setValue] = useState<number | undefined>(999);
  const onSliderChange = (v: React.SetStateAction<number>) => {
    setWidth(v);
  };
  return (
    <>
      <Space direction="vertical" size="large">
        <InputNumber
          value={value}
          formatter={e => `okokokokokokokokokokokokokok${e}okokokokokokokokokokokokokok`}
          style={{ width }}
        />
        <InputNumber
          value={value}
          formatter={e => `okokokokokokokokokokokokokok${e}okokokokokokokokokokokokokok`}
          style={{ width: 100 }}
        />
        <InputNumber
          value={value}
          formatter={e => `okokokokokokokokokokokokokok${e}okokokokokokokokokokokokokok`}
          style={{ width: 200 }}
        />
        <InputNumber
          value={value}
          formatter={e => `okokokokokokokokokokokokokok${e}okokokokokokokokokokokokokok`}
          style={{ width: 300 }}
        />
        <div />
        <Space>
          <InputNumber
            style={{ width: 300 }}
            onChange={e => {
              console.log(e);
              e ? setValue(e || 0) : setValue(undefined);
            }}
          />
          <span>输入</span>
        </Space>
        <Slider
          style={{
            width: 400,
          }}
          min={30}
          max={500}
          defaultValue={30}
          onChange={onSliderChange}
        />
      </Space>
    </>
  );
};

export const FormatterWithPrecision = () => {
  const [v, setV] = React.useState<number | undefined>();
  const onChangeHandler = (changed: number | undefined) => {
    setV(changed);
    console.log('onChange Triggered: ', changed);
  };
  return (
    <>
      <h4> Formatter / Precision / Max</h4>
      <p>Formatter 需要实时生效， precison/ max 只有 blur 或者 step 时才能生效</p>
      <p> 受控：</p>
      <InputNumber
        value={v}
        formatter={value => `$ ${value}`.replace(/\B(?=(\d{3})+(?!\d))/g, ',')}
        parser={value => value?.replace(/\$\s?|(,*)/g, '') || ''}
        onChange={onChangeHandler}
        precision={2}
        max={100}
      />
      <p> 非受控：</p>
      <InputNumber
        formatter={value => `$ ${value}`.replace(/\B(?=(\d{3})+(?!\d))/g, ',')}
        parser={value => value?.replace(/\$\s?|(,*)/g, '') || ''}
        onChange={changed => console.log('onChange Triggered: ', changed)}
        precision={2}
        max={100}
      />
      <InputNumber
        value={v}
        onChange={onChangeHandler}
        formatter={value => (Number(value) * 100).toString()}
        parser={value => String(Number(value) / 100)}
        precision={4}
        max={1}
        min={0}
        onlyNumber
        step={0.0001}
      />
    </>
  );
};
