// @universe-design/react/src/input-number/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import CAddon from '../demo/addon';
import CBasic from '../demo/basic';
import CDigit from '../demo/digit';
import CFormatter from '../demo/formatter';
import CHideControls from '../demo/hideControls';
import CMode from '../demo/mode';
import CNumbersOnly from '../demo/numbers-only';
import CSize from '../demo/size';
if (!(CAddon as Story).storyName) {
  (CAddon as Story).storyName = '3. addon';
  (CAddon as Story).parameters = {
    docs: {
      description: {
        story: '可以通过 `addonBefore` 或 `addonAfter` 来设置额外的前缀或后缀。\n\n',
      },
    },
  };
}
if (!(CBasic as Story).storyName) {
  (CBasic as Story).storyName = '0. 基础用法';
  (CBasic as Story).parameters = {
    docs: {
      description: {
        story: '简单的用法。\n\n',
      },
    },
  };
}
if (!(CDigit as Story).storyName) {
  (CDigit as Story).storyName = '2. 小数';
  (CDigit as Story).parameters = {
    docs: {
      description: {
        story: '和原生的数字输入框一样，`value`的精度由`step`的小数位数决定。\n\n',
      },
    },
  };
}
if (!(CFormatter as Story).storyName) {
  (CFormatter as Story).storyName = '4. 格式化展示';
  (CFormatter as Story).parameters = {
    docs: {
      description: {
        story: '通过`formatter`格式化数字，以展示具有具体含义的数据，往往需要配合`parser`一起使用。\n\n',
      },
    },
  };
}
if (!(CHideControls as Story).storyName) {
  (CHideControls as Story).storyName = '6. 隐藏按钮';
  (CHideControls as Story).parameters = {
    docs: {
      description: {
        story: '通过设置`controls`为`false` 可以隐藏加、减按钮。\n\n',
      },
    },
  };
}
if (!(CMode as Story).storyName) {
  (CMode as Story).storyName = '1. 步进模式';
  (CMode as Story).parameters = {
    docs: {
      description: {
        story: '通过设置`mode`为`stepper`可以使用步进模式的数字输入框。\n\n',
      },
    },
  };
}
if (!(CNumbersOnly as Story).storyName) {
  (CNumbersOnly as Story).storyName = '5. 只允许输入数字';
  (CNumbersOnly as Story).parameters = {
    docs: {
      description: {
        story: '可以通过 `onlyNumber` 属性，自动把输入的字符过滤, 只保留数字：\n\n',
      },
    },
  };
}
if (!(CSize as Story).storyName) {
  (CSize as Story).storyName = '6. 三种尺寸';
  (CSize as Story).parameters = {
    docs: {
      description: {
        story: '数字输入框有 3 种尺寸，分别为`large`、`medium`、`small`，默认尺寸为`medium`。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/InputNumber/官网示例',
  tags: ['autodocs'],
};

export { CAddon, CBasic, CDigit, CFormatter, CHideControls, CMode, CNumbersOnly, CSize };
