// @universe-design/react/src/shared/components/picker/Comp/panels/time-panel/Column.tsx
/* eslint-disable @typescript-eslint/no-magic-numbers */

import type { ListRef } from '@universe-design/rc-virtual-list';
import VList from '@universe-design/rc-virtual-list';
import stylex from '@universe-design/stylex';
import clsx from 'clsx';
import { useCssScope } from '../../../../../../config-provider';
import React, { useRef } from 'react';
import { useEffectOnce, useUpdateEffect } from '../../../../../../hooks';
import { __TEST__ } from '../../../../../../util';
import { ListItem, ListItemTitle } from '../../../../list';
import { ListBox, ListScrollView } from '../../../../list/List';
import { usePickerContextSelector } from '../../../Context';
import type { GenerateConfig } from '../../../generate';
import { scrollTo } from '../../../utils';
import { isSameTime } from '../../../utils/date';
import type { Unit } from './index';
import { cssTypographyTitle4 } from '../../../../../../stylex';

export type ColumnProps<DateType> = {
  units?: Unit<DateType>[];
  value: number | DateType;
  mode?: string;
  hideDisabledOptions?: boolean;
  onPanelChange?: (value: number | DateType) => void;
  generateConfig: GenerateConfig<DateType>;
  ulWidth?: number;
  title?: string;
};

function Column<DateType>(props: ColumnProps<DateType>) {
  const { units, onPanelChange, value, hideDisabledOptions, mode, generateConfig, ulWidth } = props;

  const inDatePicker = usePickerContextSelector(o => o!.inDatePicker);
  // 是否被嵌套在 datePicker 中
  const cssScope = useCssScope('time-panel');
  const prefixCls = cssScope();
  const itemCount = inDatePicker ? 264 / 32 : 7.5;
  const listRef = useRef<HTMLUListElement>(null);
  const vListRef = useRef<ListRef>(null);
  const liRefs = useRef<Map<number, HTMLElement | null>>(new Map());

  const scrollToItem = React.useCallback((_value: number, duration = 120) => {
    const li = liRefs.current.get(_value);
    const list = listRef.current;
    if (li && list) {
      const LIST_ITEM_MARGIN_TOP = 1; // 第一个元素 margin-top: 1;
      scrollTo(list, li.offsetTop - LIST_ITEM_MARGIN_TOP, duration);
    }
    const vList = vListRef.current;
    if (vList) {
      vList.scrollTo({ key: _value, align: 'top' });
    }
  }, []);

  useEffectOnce(() => {
    // TODO 这里不用异步会获取不到 li 的尺寸和位置，后面需要看一下
    requestAnimationFrame(() => {
      scrollToItem(Number(value), 0);
    });
  });

  useUpdateEffect(
    React.useCallback(() => {
      scrollToItem(Number(value));
    }, [scrollToItem, value]),
  );

  const isSameTimeDateType = (time1: number | DateType, time2: number | DateType): boolean => {
    if (mode === 'timeCombine') {
      return time1 !== -1 ? isSameTime(generateConfig, time1 as DateType, time2 as DateType) : false;
    } else {
      return false;
    }
  };

  const isSame = (time1: number | DateType, time2: number | DateType): boolean => {
    if (mode === 'timeCombine') {
      return isSameTimeDateType(time1, time2);
    }
    return time1 === time2;
  };

  const renderItem = (unit: Unit<DateType>) => {
    if (hideDisabledOptions && unit.disabled) {
      return null;
    }
    return (
      <ListItem
        key={Number(unit.value)}
        ref={element => {
          liRefs.current.set(Number(unit.value), element);
        }}
        onClick={() => {
          if (unit.disabled) {
            return;
          }
          onPanelChange && onPanelChange(unit.value);
        }}
        disabled={unit.disabled}
        selected={isSame(value, unit.value)}
        showCheckmark={mode === 'timeCombine'}
      >
        <ListItemTitle>{unit.label}</ListItemTitle>
      </ListItem>
    );
  };

  const renderUl = () =>
    mode === 'timeCombine' ? (
      <VList<Unit<DateType>>
        data={units ?? []}
        itemKey={unit => Number(unit.value)}
        fullHeight={false}
        itemHeight={32}
        height={32 * itemCount}
        virtual={!__TEST__}
        ref={vListRef}
        style={{
          width: '100%',
        }}
      >
        {unit => {
          const item = renderItem(unit);
          if (item) {
            return <div>{item}</div>;
          }
        }}
      </VList>
    ) : (
      <ListBox className={`${prefixCls}__column`}>
        {inDatePicker && (
          <span
            className={clsx(`${prefixCls}__column-title`, stylex(cssTypographyTitle4))}
            data-testid="time-column-title"
          >
            {props.title}
          </span>
        )}
        <ListScrollView
          style={{
            width: ulWidth,
            height: itemCount * 32,
            position: 'relative', // ele.scrollTop is offset of closest relative positioned parent.
          }}
          className={`${prefixCls}__column-scroll-list`}
          ref={listRef}
        >
          {(units ?? []).map(renderItem)}
          <div
            style={{
              paddingBottom: (itemCount - 1) * 32,
            }}
          />
        </ListScrollView>
      </ListBox>
    );

  return renderUl();
}

export default Column;
