// @universe-design/react/src/shared/components/picker/Comp/panels/time-panel/CombineTimePanel.tsx
/* eslint-disable @typescript-eslint/no-magic-numbers */
import React from 'react';
import { useLocale } from '../../../../../../config-provider/hooks';
import Column from './Column';
import { leftPad } from '../../../utils';
import { parseValue } from '../../../utils/ui';
import type { CombineTimePanelProps } from '../../../interface';
import type { Unit } from './index';
import { isSameTime } from '../../../utils/date';
import { ListBox } from '../../../../list/List';

function CombineTimePanel<DateType>(props: CombineTimePanelProps<DateType>) {
  const {
    prefixCls,
    activeColumnIndex,
    value = -1,
    use12Hours,
    timeStep = 15, // 时间间隔，单位是分钟，默认15min，建议15min, 30min, 45min, 60min
    disabledTime,
    generateConfig,
    parse,
    mode,
    locale,
    hideDisabledOptions,
    onPanelChange,
    customTime = [],
    popupStyle,
  } = props;

  const { locale: localeName } = useLocale();

  // timeCombine格式下将其统一化为HH:mm格式比较
  const option = {
    format: 'HH:mm',
    parse,
    mode,
    locale,
    localeName,
  };

  const parseWidth = popupStyle && parseFloat(popupStyle?.width as string);

  const validTimeSteps = (number: number) => {
    if (60 % number === 0) {
      return true;
    } else {
      return false;
    }
  };

  const columns: {
    node: React.ReactElement;
    value: DateType | -1;
    units: Unit<DateType>[];
    onPanelChange: (diff: DateType) => void;
  }[] = [];
  const columnPrefixCls = `${prefixCls}-time-panel`;

  const time = value ? value : -1;

  function generateUnits(
    start: number,
    end: number,
    step: number, // 分钟
    disabledUnits: (DateType | null)[],
  ) {
    const units: Unit<DateType>[] = [];
    if (step < 60) {
      for (let i = start; i <= end; i += 1) {
        for (let j = 0; j <= 59; j += step) {
          const label = `${i > 9 ? i : `0${i}`}:${leftPad(j, 2)}`;
          const parseTime = parseValue<DateType>(generateConfig, label, option);
          const isDisabled =
            !disabledUnits?.some(item => isSameTime<DateType>(generateConfig, item, parseTime)) || false;

          units.push({
            label,
            value: parseTime || -1,
            disabled: !isDisabled,
          });
        }
      }
    } else if (step === 60) {
      for (let i = start; i <= end; i += 1) {
        const label = i > 9 ? `${i}:00` : `0${i}:00`;
        const parseTime = parseValue<DateType>(generateConfig, label, option);
        const isDisabled = !disabledUnits?.some(item => isSameTime(generateConfig, item, parseTime)) || false;
        units.push({
          label,
          value: parseTime || -1,
          disabled: !isDisabled,
        });
      }
    }
    return units;
  }

  const disabledTimes = disabledTime?.map((i: string) => parseValue(generateConfig, i, option)) || [];

  // 在这里可以通过修改前两个参数从而改变时间的开始与结束 customTime自定义的时间列表
  const rawTime =
    customTime.length > 0 ? customTime : generateUnits(0, 23, validTimeSteps(timeStep) ? timeStep : 15, disabledTimes);

  const showTime = hideDisabledOptions ? rawTime.filter(i => !i.disabled) : rawTime;

  const memorizedRawTime = showTime;

  const times = React.useMemo(() => {
    if (!use12Hours) {
      return memorizedRawTime;
    }
    return memorizedRawTime.map(v => ({
      ...v,
      label: generateConfig.formatLabel(v.value, 'hh:mm A'),
    }));
  }, [use12Hours, memorizedRawTime, generateConfig]);

  // ======================== Render ========================
  function addColumnNode(
    node: React.ReactElement,
    columnValue: DateType | -1,
    units: Unit<DateType>[],
    onColumnSelect: (diff: DateType) => void,
  ) {
    columns.push({
      node: React.cloneElement(node, {
        prefixCls: columnPrefixCls,
        value: columnValue,
        active: activeColumnIndex === columns.length,
        onPanelChange: onColumnSelect,
        units,
        hideDisabledOptions,
      }),
      onPanelChange: onColumnSelect,
      value: columnValue,
      units,
    });
  }

  addColumnNode(
    <Column<DateType> key="timeCombine" mode="timeCombine" value={value} generateConfig={generateConfig} />,
    time,
    times,
    num => onPanelChange(num),
  );

  const panelPrefixCls = `${prefixCls}-time-combine-panel`;

  return (
    <div className={panelPrefixCls}>
      <ListBox style={{ width: parseWidth ? parseWidth : 144 }}>{columns.map(({ node }) => node)}</ListBox>
    </div>
  );
}

export default CombineTimePanel;
