// @universe-design/react/src/shared/components/picker/Comp/panels/time-panel/index.tsx
import React from 'react';
import type { TimePanelProps } from '../../../interface';
import { createKeyDownHandler } from '../../../utils';
import NormalTimePanel from './NormalTimePanel';
import CombineTimePanel from './CombineTimePanel';
import { generateFormat } from '../../../utils/date';

export type BodyOperationRef = {
  onUpDown: (diff: number) => void;
};

export type Unit<DateType> = {
  label: React.ReactText;
  value: DateType | number;
  disabled: boolean;
};

const countBoolean = (boolList: (boolean | undefined)[]) => boolList.filter(bool => bool !== false).length;

export default function TimePanel<DateType>(props: TimePanelProps<DateType>) {
  const {
    generateConfig,
    prefixCls,
    operationRef,
    value,
    format,
    onPanelChange,
    mode = 'time',
    showSecond: showSecondProp,
    showMinute: showMinuteProp,
    showHour: showHourProp,
    use12Hours: use12HoursProp,
  } = props;

  const showSecond = showSecondProp ?? (format ? generateFormat.showSecond(format) : false);
  const showMinute = showMinuteProp ?? (format ? generateFormat.showMinute(format) : true);
  const showHour = showHourProp ?? (format ? generateFormat.showHour(format) : true);
  const use12Hours = use12HoursProp ?? (format ? generateFormat.is12Hour(format) : false);
  const panelPrefixCls = `${prefixCls}-time-panel`;
  const bodyOperationRef = React.useRef<BodyOperationRef>();

  // ======================= Keyboard =======================
  const [activeColumnIndex, setActiveColumnIndex] = React.useState(-1);
  const columnsCount = countBoolean([showSecond, use12Hours]);

  operationRef &&
    (operationRef.current = {
      onKeyDown: event =>
        createKeyDownHandler(event, {
          onLeftRight: (diff: number) => {
            setActiveColumnIndex((activeColumnIndex + diff + columnsCount) % columnsCount);
          },
          onUpDown: (diff: number) => {
            if (activeColumnIndex === -1) {
              setActiveColumnIndex(0);
            } else if (bodyOperationRef.current) {
              bodyOperationRef.current.onUpDown(diff);
            }
          },
          onEnter: () => {
            onPanelChange(value || generateConfig.getNow());
            setActiveColumnIndex(-1);
          },
        }),

      onBlur: () => {
        setActiveColumnIndex(-1);
      },
    });

  const renderPanel = () => {
    if (mode === 'timeCombine') {
      return <CombineTimePanel {...props} use12Hours={use12Hours} />;
    } else if (mode === 'time') {
      return (
        <div className={panelPrefixCls}>
          <NormalTimePanel
            {...props}
            showMinute={showMinute}
            showHour={showHour}
            use12Hours={use12Hours}
            showSecond={showSecond}
            activeColumnIndex={activeColumnIndex}
          />
        </div>
      );
    }
  };

  return <div>{renderPanel()}</div>;
}
