// @universe-design/react/src/shared/components/picker/Comp/panels/time-panel/NormalTimePanel.tsx
/* eslint-disable @typescript-eslint/no-magic-numbers */
import { useCssScope } from '../../../../../../config-provider';
import { useDatePickerContext } from '../../../../../../date-picker/Context';
import clsx from 'clsx';
import React from 'react';
import { ListBox } from '../../../../list/List';
import { usePickerContext } from '../../../Context';
import type { NormalTimePanelProps } from '../../../interface';
import { leftPad, setTime as utilSetTime, useMemo } from '../../../utils';
import Column from './Column';
import type { Unit } from './index';
import './style';

function shouldUnitsUpdate(prevUnits: Unit<number>[], nextUnits: Unit<number>[]) {
  if (prevUnits.length !== nextUnits.length) {
    return true;
  }
  // if any unit's disabled status is different, the units should be re-evaluted
  for (let i = 0; i < prevUnits.length; i += 1) {
    if (prevUnits[i].disabled !== nextUnits[i].disabled) {
      return true;
    }
  }
  return false;
}

function generateUnits(start: number, end: number, step: number, disabledUnits: number[] | undefined) {
  const units: Unit<number>[] = [];
  for (let i = start; i <= end; i += step) {
    units.push({
      label: leftPad(i, 2),
      value: i,
      disabled: (disabledUnits || []).includes(i),
    });
  }
  return units;
}

function NormalTimePanel<DateType>(props: NormalTimePanelProps<DateType>) {
  const {
    generateConfig,
    prefixCls,
    activeColumnIndex,
    showSecond = false,
    showMinute = true,
    showHour = true,
    use12Hours,
    hourStep = 1,
    minuteStep = 1,
    secondStep = 1,
    disabledHours,
    disabledMinutes,
    disabledSeconds,
    hideDisabledOptions,
    onPanelChange,
    locale,
    popupStyle,
  } = props;

  const { value: valueFromCtx } = usePickerContext<DateType>();
  const datePickerContext = useDatePickerContext();
  const cssScope = useCssScope('time-panel');
  const classNames = clsx(datePickerContext.showTime && cssScope('--in-datepicker'));

  const value = props.value ?? valueFromCtx ?? -1;
  const parseWidth = popupStyle && parseFloat(popupStyle?.width as string);
  const rowLength = () => {
    const rowArray = [showSecond, showHour, showMinute, use12Hours];
    return rowArray.filter(item => item).length;
  };

  const ulWidth = parseWidth ? Math.floor(parseWidth / rowLength()) : 70;

  const columns: {
    node: React.ReactElement;
    value: number;
    units: Unit<number>[];
    onPanelChange: (diff: number) => void;
  }[] = [];
  const columnPrefixCls = `${prefixCls}-time-panel`;

  let isPM = false;
  const originHour = value === -1 ? -1 : generateConfig.getHour(value);
  let hour = originHour;
  const minute = value === -1 ? -1 : generateConfig.getMinute(value);
  const second = value === -1 ? -1 : generateConfig.getSecond(value);

  const setTime = (isNewPM: boolean, newHour: number, newMinute: number, newSecond: number) => {
    let newDate = value === -1 ? generateConfig.getNow() : value;

    const mergedHour = Math.max(0, newHour);
    const mergedMinute = Math.max(0, newMinute);
    const mergedSecond = Math.max(0, newSecond);

    newDate = utilSetTime(
      generateConfig,
      newDate,
      !use12Hours || !isNewPM ? mergedHour : mergedHour + 12,
      mergedMinute,
      mergedSecond,
    );

    return newDate;
  };

  // ========================= Unit =========================
  const rawHours = generateUnits(0, 23, hourStep, disabledHours && disabledHours());

  const memorizedRawHours = useMemo(
    () => rawHours,
    // eslint-disable-next-line react-hooks/exhaustive-deps
    rawHours,
    shouldUnitsUpdate,
  );

  // Should additional logic to handle 12 hours
  if (use12Hours && hour) {
    isPM = hour >= 12; // -1 means should display AM
    hour %= 12;
  }

  const hours = React.useMemo(() => {
    if (!use12Hours) {
      return memorizedRawHours || [];
    }
    if (memorizedRawHours) {
      return (
        memorizedRawHours
          .filter(isPM ? hourMeta => hourMeta.value >= 12 : hourMeta => hourMeta.value < 12)
          .map(hourMeta => {
            const hourValue = hourMeta.value % 12;
            const hourLabel = hourValue === 0 ? '12' : leftPad(hourValue, 2);

            return {
              ...hourMeta,
              label: hourLabel,
              value: hourValue,
            };
          }) || []
      );
    }
    return [];
  }, [use12Hours, isPM, memorizedRawHours]);

  const minutes = generateUnits(0, 59, minuteStep, disabledMinutes && disabledMinutes(originHour));

  const seconds = generateUnits(0, 59, secondStep, disabledSeconds && disabledSeconds(originHour, minute));

  // ======================== Render =======================
  function addColumnNode(
    node: React.ReactElement,
    columnValue: number,
    units: Unit<number>[],
    onColumnSelect: (diff: number) => void,
  ) {
    columns.push({
      node: React.cloneElement(node, {
        prefixCls: columnPrefixCls,
        value: columnValue,
        active: activeColumnIndex === columns.length,
        onPanelChange: onColumnSelect,
        units,
        hideDisabledOptions,
      }),
      onPanelChange: onColumnSelect,
      value: columnValue,
      units,
    });
  }

  use12Hours &&
    addColumnNode(
      <Column<DateType> key="12hours" value={value} ulWidth={ulWidth} generateConfig={generateConfig} />,
      // 0是上午1是下午
      isPM ? 1 : 0,
      [
        { label: locale.morning, value: 0, disabled: false },
        { label: locale.afternoon, value: 1, disabled: false },
      ],
      num => {
        onPanelChange(setTime(Boolean(num), hour, minute, second));
      },
    );

  showHour &&
    addColumnNode(
      <Column<DateType>
        key="hour"
        ulWidth={ulWidth}
        value={value}
        generateConfig={generateConfig}
        title={locale.hour}
      />,
      hour,
      hours,
      num => {
        onPanelChange(setTime(isPM, num, minute, second));
      },
    );

  showMinute &&
    addColumnNode(
      <Column<DateType>
        key="minute"
        value={value}
        ulWidth={ulWidth}
        generateConfig={generateConfig}
        title={locale.minute}
      />,
      minute,
      minutes,
      num => {
        onPanelChange(setTime(isPM, hour, num, second));
      },
    );

  showSecond &&
    addColumnNode(
      <Column<DateType>
        key="second"
        ulWidth={ulWidth}
        value={value}
        generateConfig={generateConfig}
        title={locale.second}
      />,
      second,
      seconds,
      num => {
        onPanelChange(setTime(isPM, hour, minute, num));
      },
    );

  return (
    <ListBox cascaded className={classNames}>
      {columns.map(({ node }) => node)}
    </ListBox>
  );
}

export default NormalTimePanel;
