// @universe-design/react/lib/image-viewer/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WHEEL_DELAY = exports.VEC2_SCALE_OF_ORIGINAL_SIZE = exports.TO_FIXED_NUMBER = exports.TOOLBAR_FALSE = exports.TOOLBAR_DEFAULT = exports.SCALING_SPEED = exports.SCALE_OF_ORIGINAL_SIZE = exports.ROTATION_RADIAN = exports.PLUGIN_NUM_2 = exports.PLUGIN_NUM = exports.MIN_SCALE_NUM = exports.MAX_SCALE_NUM = exports.FIT_TO_CONTAINER_BOUNDARY_PADDING = exports.DELTA_Y_SCALE = exports.CURRENT_SCALE_OF_ORIGINAL_SIZE = exports.AMPLIFY_NUMBER = void 0;
// 缩放递增常量
const SCALING_SPEED = exports.SCALING_SPEED = 0.15;
// 最小缩放
const MIN_SCALE_NUM = exports.MIN_SCALE_NUM = 0.1;
// 最大缩放
const MAX_SCALE_NUM = exports.MAX_SCALE_NUM = 5;
// 原始尺寸的比例
const FIT_TO_CONTAINER_BOUNDARY_PADDING = exports.FIT_TO_CONTAINER_BOUNDARY_PADDING = 72;
const SCALE_OF_ORIGINAL_SIZE = exports.SCALE_OF_ORIGINAL_SIZE = 0.87;
const VEC2_SCALE_OF_ORIGINAL_SIZE = exports.VEC2_SCALE_OF_ORIGINAL_SIZE = 0.5;
// getNextScale
const CURRENT_SCALE_OF_ORIGINAL_SIZE = exports.CURRENT_SCALE_OF_ORIGINAL_SIZE = 1;
const DELTA_Y_SCALE = exports.DELTA_Y_SCALE = 0.01;
// 选择角度
const ROTATION_RADIAN = exports.ROTATION_RADIAN = -Math.PI / 2;
// 保留小数位数
const TO_FIXED_NUMBER = exports.TO_FIXED_NUMBER = 2;
// 小数放大
const AMPLIFY_NUMBER = exports.AMPLIFY_NUMBER = 100;
// 初始需求插件总数
const PLUGIN_NUM = exports.PLUGIN_NUM = 4;
// 当插件个数为2
const PLUGIN_NUM_2 = exports.PLUGIN_NUM_2 = 2;
// 触摸板移动图片的延时
const WHEEL_DELAY = exports.WHEEL_DELAY = 300;
const TOOLBAR_DEFAULT = exports.TOOLBAR_DEFAULT = {
  pagination: true,
  scaling: true,
  rotation: true,
  download: true
};
const TOOLBAR_FALSE = exports.TOOLBAR_FALSE = {
  pagination: false,
  scaling: false,
  rotation: false,
  download: false
};
