// @universe-design/react/lib/image-viewer/ImageViewerDialog.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ImageViewerDialog = void 0;
var _u13d3rh = _interopRequireDefault(require("./../.stylex/u13d3rh8"));
var _udg2fzc = _interopRequireDefault(require("./../.stylex/udg2fzc"));
var _u1u1p7mj = _interopRequireDefault(require("./../.stylex/u1u1p7mj"));
var _react = require("@universe-design/stylex/react");
var React = _interopRequireWildcard(require("react"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _dialog = require("../dialog/dialog");
var _useImageViewerVisible = require("./useImageViewerVisible");
var _button = require("../button");
var _CloseOutlined = _interopRequireDefault(require("@universe-design/icons-react/CloseOutlined"));
var _ImageViewerContext = require("./ImageViewerContext");
require("./style");
//.uly5x3f{width:100%;}
//.u1l02sjl{height:100%;}
//.u1euv43f{position:absolute;}
/* eslint-disable react/no-unknown-property */
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const ImageViewerDialog = exports.ImageViewerDialog = /*#__PURE__*/React.forwardRef((props, ref) => {
  // TODO 原则上来说这块也可以有一个 useImageViewerDialogProps 的 hooks 专门预处理 props，我们的 render 函数体里要尽可能干净
  const {
    zIndex,
    maskClosable = true,
    keyboardClosable,
    hasCloseButton = true,
    children
  } = props;
  const {
    rootCssScope,
    containerRef,
    dialogVisible,
    setDialogVisible
  } = (0, _ImageViewerContext.useImageViewerContext)();
  return (0, _react.jsx)(_dialog.DialogLayoutMode, {
    mode: "2.18.0"
  }, (0, _react.jsx)(_dialog.Dialog, {
    ref: ref,
    zIndex: zIndex,
    safeMargin: {
      hmax: 0,
      hmin: 0,
      top: 0,
      bottom: 0
    },
    width: "100%",
    wrapperClassName: rootCssScope('__dialog__wrap'),
    destroyOnClose: true,
    keyboardClosable: keyboardClosable,
    overlayClosable: maskClosable,
    overlayClassName: rootCssScope('__dialog__mask'),
    visible: dialogVisible
    // dialog mask 和 esc 关闭
    ,
    onClose: (e, _ref) => {
      let {
        cancelType
      } = _ref;
      if (cancelType === _dialog.DialogCancelType.Esc) {
        setDialogVisible(false, {
          visible: false,
          event: e,
          action: _useImageViewerVisible.ImageViewerCloseAction.Esc
        });
        return;
      }
      if (cancelType === _dialog.DialogCancelType.Mask) {
        setDialogVisible(false, {
          visible: false,
          event: e,
          action: _useImageViewerVisible.ImageViewerCloseAction.Mask
        });
      }
    }
  }, (0, _react.jsx)("div", {
    ref: containerRef,
    sx: [{
      "props": {
        "a9b677": _u13d3rh.default,
        "Bqenvij": _udg2fzc.default,
        "qhf8xq": _u1u1p7mj.default
      },
      "alias": rootCssScope()
    }]
  }, children, hasCloseButton && (0, _react.jsx)(_button.IconButton, {
    size: 20,
    icon: (0, _react.jsx)(_CloseOutlined.default, null),
    className: rootCssScope('__close-button'),
    onClick: ev => {
      setDialogVisible(false, {
        visible: false,
        event: ev,
        action: _useImageViewerVisible.ImageViewerCloseAction.CloseButton
      });
    }
  }))));
});
