// @universe-design/react/lib/image-viewer/ImageViewer.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ImageViewer = void 0;
var _react = require("@universe-design/stylex/react");
var React = _interopRequireWildcard(require("react"));
var _ImageViewerToolbar = require("./ImageViewerToolbar");
var _ImageViewerContext = require("./ImageViewerContext");
var _ImageViewerDialog = require("./ImageViewerDialog");
var _ImageViewerImage = require("./ImageViewerImage");
var _useImageViewer = require("./useImageViewer");
var _slot = require("../shared/slot");
var _constants = require("./constants");
require("./style");
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars

const ImageViewer = props => {
  const {
    errorFallback,
    toolbar: toolbarProp = true,
    plugins
  } = props;
  const context = (0, _useImageViewer.useImageViewer)(props);
  const toolbarSlot = (0, _slot.resolveSlot)((() => {
    if (toolbarProp === true) {
      return _constants.TOOLBAR_DEFAULT;
    }
    if (toolbarProp === false) {
      return _constants.TOOLBAR_FALSE;
    }
    return toolbarProp;
  })(), {
    elementType: _ImageViewerToolbar.ImageViewerToolbar,
    defaultProps: {
      pagination: _constants.TOOLBAR_DEFAULT.pagination,
      scaling: _constants.TOOLBAR_DEFAULT.scaling,
      rotation: _constants.TOOLBAR_DEFAULT.rotation,
      download: _constants.TOOLBAR_DEFAULT.download
    },
    renderByDefault: true
  });
  return (0, _react.jsx)(_ImageViewerContext.ImageViewerContextProvider, {
    value: context
  }, (0, _react.jsx)(_ImageViewerDialog.ImageViewerDialog, props, (0, _react.jsx)(_ImageViewerImage.ImageViewerImage, {
    errorFallback
  }), (0, _slot.renderSlotNode)(toolbarSlot, plugins)));
};
exports.ImageViewer = ImageViewer;
