// @universe-design/react/lib/image-viewer/ImageViewerToolbar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ImageViewerToolbar = void 0;
var _utkiswu = _interopRequireDefault(require("./../.stylex/utkiswu"));
var _u1ny05fg = _interopRequireDefault(require("./../.stylex/u1ny05fg"));
var _u256sz = _interopRequireDefault(require("./../.stylex/u256sz"));
var _u1443mj = _interopRequireDefault(require("./../.stylex/u1443mj0"));
var _u1u1p7mj = _interopRequireDefault(require("./../.stylex/u1u1p7mj"));
var _u1l13cd = _interopRequireDefault(require("./../.stylex/u1l13cd8"));
var _ukchsq = _interopRequireDefault(require("./../.stylex/ukchsq7"));
var _ubyhozq = _interopRequireDefault(require("./../.stylex/ubyhozq"));
var _ut7q8es = _interopRequireDefault(require("./../.stylex/ut7q8es"));
var _us2xlyu = _interopRequireDefault(require("./../.stylex/us2xlyu"));
var _ukjz3g = _interopRequireDefault(require("./../.stylex/ukjz3g"));
var _u19bk9ok = _interopRequireDefault(require("./../.stylex/u19bk9ok"));
var _uathu2q = _interopRequireDefault(require("./../.stylex/uathu2q"));
var _urnexrx = _interopRequireDefault(require("./../.stylex/urnexrx"));
var _u12hcvx = _interopRequireDefault(require("./../.stylex/u12hcvx2"));
var _u1gja64x = _interopRequireDefault(require("./../.stylex/u1gja64x"));
var _u162mx8s = _interopRequireDefault(require("./../.stylex/u162mx8s"));
var _u1jmni = _interopRequireDefault(require("./../.stylex/u1jmni06"));
var _u9yw1hi = _interopRequireDefault(require("./../.stylex/u9yw1hi"));
var _u1ksa1gg = _interopRequireDefault(require("./../.stylex/u1ksa1gg"));
var _u1fdqv2r = _interopRequireDefault(require("./../.stylex/u1fdqv2r"));
var _u143d86t = _interopRequireDefault(require("./../.stylex/u143d86t"));
var _u6qvdto = _interopRequireDefault(require("./../.stylex/u6qvdto"));
var _urjm3ap = _interopRequireDefault(require("./../.stylex/urjm3ap"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _react2 = require("@universe-design/stylex/react");
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _ImageViewerContext = require("./ImageViewerContext");
var _forwardRef = require("../util/forwardRef");
var _slot = require("../shared/slot");
var _space = require("../space");
var _Pagination = require("./components/Pagination");
var _Scaling = require("./components/Scaling");
var _Rotation = require("./components/Rotation");
var _Download = require("./components/Download");
require("./style");
//.urkrog8{width:1px;}
//.ujamq6b{height:20px;}
//.u11fluvs{background:@udtoken-dropdown-button-divider;}
//.u1xqy1su{user-select:none;}
//.u1euv43f{position:absolute;}
//.u10cq58m{bottom:16px;}
//.u22iagw{display:flex;}
//.u122n59{align-items:center;}
//.u4d9j23{justify-content:center;}
//.uf2sm71{height:48px;}
//.u1ahpp82{width:max-content;}
//.u198b2fy{background-color:@udtoken-imageview-toolbox-bg;}
//.u19ibtj8{color:@static-white;}
//.ugvjd9r{border-top-left-radius:@border-radius-xs;}
//.uns275c{border-top-right-radius:@border-radius-xs;}
//.udbqa1p{border-bottom-right-radius:@border-radius-xs;}
//.u1vdvwtp{border-bottom-left-radius:@border-radius-xs;}
//.u1ktbn1t{left:50%;}
//.u13gfj74{transform:translateX(-50%);}
//.u1cguypg{pointer-events:auto;}
//.u1g0x7ka{padding-top:0;}
//.u11qrl6u{padding-right:12px;}
//.u1qch9an{padding-bottom:0;}
//.ujlbh76{padding-left:12px;}
/* eslint-disable react/no-unknown-property */
/* eslint-disable react/jsx-no-leaked-render */
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const useImageViewerToolbarStylexStyles = (0, _react2.makeStylexStyles)(_ref => {
  let {
    prefix
  } = _ref;
  return {
    toolbar: {
      "props": {
        "famaaq": _u1443mj.default,
        "qhf8xq": _u1u1p7mj.default,
        "B5kzvoi": _u1l13cd.default,
        "mc9l5x": _ukchsq.default,
        "Bt984gj": _ubyhozq.default,
        "Brf1p80": _ut7q8es.default,
        "Bqenvij": _us2xlyu.default,
        "a9b677": _ukjz3g.default,
        "De3pzq": _u19bk9ok.default,
        "sj55zd": _uathu2q.default,
        "Btl43ni": _urnexrx.default,
        "B7oj6ja": _u12hcvx.default,
        "Bbmb7ep": _u1gja64x.default,
        "Beyfa6y": _u162mx8s.default,
        "oyh7mz": _u1jmni.default,
        "Bz10aip": _u9yw1hi.default,
        "Bkecrkj": _u1ksa1gg.default,
        "z8tnut": _u1fdqv2r.default,
        "z189sj": _u143d86t.default,
        "Byoj8tv": _u6qvdto.default,
        "uwmqm3": _urjm3ap.default
      },
      "alias": prefix
    },
    toolbarDivider: {
      "props": {
        "a9b677": _utkiswu.default,
        "Bqenvij": _u1ny05fg.default,
        "ayd6f0": _u256sz.default
      },
      "alias": `${prefix}__divider`
    }
  };
});
const ImageViewerToolbar = exports.ImageViewerToolbar = (0, _forwardRef.forwardHTMLRef)((props, ref) => {
  const {
    pagination,
    scaling,
    rotation,
    download: downloadProp,
    children,
    ...htmlProps
  } = props;
  const {
    rootCssScope
  } = (0, _ImageViewerContext.useImageViewerContext)();
  const styles = useImageViewerToolbarStylexStyles({
    prefix: rootCssScope('__toolbar')
  });
  const downloadNode = (0, _slot.renderSlotNode)((0, _slot.resolveSlot)(downloadProp === true ? {} : downloadProp === false ? null : downloadProp, {
    elementType: _Download.Download,
    renderByDefault: true
  }));
  return (0, _react2.jsx)("div", (0, _extends2.default)({
    sx: [styles.toolbar]
  }, htmlProps, {
    ref: ref
  }), (0, _react2.jsx)(_space.Space, {
    size: 24,
    split: (0, _react2.jsx)("div", {
      className: (0, _stylex.default)(styles.toolbarDivider)
    })
  }, pagination && (0, _react2.jsx)(_Pagination.Pagination, null), scaling && (0, _react2.jsx)(_Scaling.Scaling, null), (rotation || downloadNode) && (0, _react2.jsx)(_space.Space, {
    size: 12
  }, rotation && (0, _react2.jsx)(_Rotation.Rotation, null), downloadNode), children));
});
