// @universe-design/react/lib/image-viewer/ImageViewerImage.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ImageViewerImage = void 0;
var _usfh9mg = _interopRequireDefault(require("./../.stylex/usfh9mg"));
var _u7iz6qb = _interopRequireDefault(require("./../.stylex/u7iz6qb"));
var _ugj19bf = _interopRequireDefault(require("./../.stylex/ugj19bf"));
var _u1it5fv = _interopRequireDefault(require("./../.stylex/u1it5fv0"));
var _u1b894eo = _interopRequireDefault(require("./../.stylex/u1b894eo"));
var _uhiotwf = _interopRequireDefault(require("./../.stylex/uhiotwf"));
var _u1cpcpb = _interopRequireDefault(require("./../.stylex/u1cpcpb5"));
var _u10qn3go = _interopRequireDefault(require("./../.stylex/u10qn3go"));
var _uflkg = _interopRequireDefault(require("./../.stylex/uflkg16"));
var _uvn5txf = _interopRequireDefault(require("./../.stylex/uvn5txf"));
var _us9j13g = _interopRequireDefault(require("./../.stylex/us9j13g"));
var _umt7kug = _interopRequireDefault(require("./../.stylex/umt7kug"));
var _uh3lwu = _interopRequireDefault(require("./../.stylex/uh3lwu4"));
var _ur = _interopRequireDefault(require("./../.stylex/ur61548"));
var _u1n7c7ic = _interopRequireDefault(require("./../.stylex/u1n7c7ic"));
var _ujkvyav = _interopRequireDefault(require("./../.stylex/ujkvyav"));
var _u182xfmz = _interopRequireDefault(require("./../.stylex/u182xfmz"));
var _uqbl7tl = _interopRequireDefault(require("./../.stylex/uqbl7tl"));
var _urnexrx = _interopRequireDefault(require("./../.stylex/urnexrx"));
var _u12hcvx = _interopRequireDefault(require("./../.stylex/u12hcvx2"));
var _u1gja64x = _interopRequireDefault(require("./../.stylex/u1gja64x"));
var _u162mx8s = _interopRequireDefault(require("./../.stylex/u162mx8s"));
var _u1u1p7mj = _interopRequireDefault(require("./../.stylex/u1u1p7mj"));
var _ufibavv = _interopRequireDefault(require("./../.stylex/ufibavv"));
var _u1jmni = _interopRequireDefault(require("./../.stylex/u1jmni06"));
var _uevjesf = _interopRequireDefault(require("./../.stylex/uevjesf"));
var _ukchsq = _interopRequireDefault(require("./../.stylex/ukchsq7"));
var _ut7q8es = _interopRequireDefault(require("./../.stylex/ut7q8es"));
var _ubyhozq = _interopRequireDefault(require("./../.stylex/ubyhozq"));
var _uh6ryxt = _interopRequireDefault(require("./../.stylex/uh6ryxt"));
var _u1ksa1gg = _interopRequireDefault(require("./../.stylex/u1ksa1gg"));
var _u1443mj = _interopRequireDefault(require("./../.stylex/u1443mj0"));
var _uit42qe = _interopRequireDefault(require("./../.stylex/uit42qe"));
var _uc4lfdq = _interopRequireDefault(require("./../.stylex/uc4lfdq"));
var _u1tngplc = _interopRequireDefault(require("./../.stylex/u1tngplc"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = require("@universe-design/stylex/react");
var React = _interopRequireWildcard(require("react"));
var _reactUtilities = require("@universe-design/react-utilities");
var _Image = require("../shared/components/Image");
var _ImageViewerContext = require("./ImageViewerContext");
var _LoadfailFilled = _interopRequireDefault(require("@universe-design/icons-react/LoadfailFilled"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _slot = require("../util/slot");
var _function = require("../util/function");
var _loading = require("../loading");
var _glMatrix = require("gl-matrix");
var _usePointerHandler = require("./usePointerHandler");
var _constants = require("./constants");
var _useImageViewerMobileScaling = require("./useImageViewerMobileScaling");
var _useImageViewerVisible = require("./useImageViewerVisible");
//.uk73vx1{opacity:0;}
//.u1825aan{color:@N400;}
//.u18m8u13{font-size:48px;}
//.uwfqyga{width:120px;}
//.ujr5b71{height:120px;}
//.u1t7pc3r{background:@N100;}
//.u5ogflp{border-top-width:1px;}
//.u1hqa2wf{border-right-width:1px;}
//.u1f09k3d{border-bottom-width:1px;}
//.uinvdd3{border-left-width:1px;}
//.uzkkow9{border-top-style:solid;}
//.ucdblym{border-right-style:solid;}
//.ug706s2{border-bottom-style:solid;}
//.ujik90z{border-left-style:solid;}
//.uyxjys6{border-top-color:@line-border-card;}
//.u1f4hk3w{border-right-color:@line-border-card;}
//.u1x626fe{border-bottom-color:@line-border-card;}
//.u14wfv0u{border-left-color:@line-border-card;}
//.ugvjd9r{border-top-left-radius:@border-radius-xs;}
//.uns275c{border-top-right-radius:@border-radius-xs;}
//.udbqa1p{border-bottom-right-radius:@border-radius-xs;}
//.u1vdvwtp{border-bottom-left-radius:@border-radius-xs;}
//.u1euv43f{position:absolute;}
//.u1i1t8d1{top:50%;}
//.u1ktbn1t{left:50%;}
//.uoouk8k{transform:translate(-50%, -50%);}
//.u22iagw{display:flex;}
//.u4d9j23{justify-content:center;}
//.u122n59{align-items:center;}
//.ujseox{display:none;}
//.u1cguypg{pointer-events:auto;}
//.u1xqy1su{user-select:none;}
//.u1sdsnyy{touch-action:none;}
//.u1b6n930{transform-origin:left top;}
//.u17vrrt3{-webkit-user-drag:none;}
/* eslint-disable react/no-unknown-property */
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const useImageViewerImageStylexStyles = (0, _react.makeStylexStyles)(_ref => {
  let {
    prefix
  } = _ref;
  return {
    image: {
      "props": {
        "Bkecrkj": _u1ksa1gg.default,
        "famaaq": _u1443mj.default,
        "lpbzjs": _uit42qe.default,
        "B76gyc8": _uc4lfdq.default,
        "Ijiw13": _u1tngplc.default
      },
      "alias": prefix
    },
    imageError: {
      "props": {
        "mc9l5x": _uh6ryxt.default
      },
      "alias": `${prefix}-error`
    },
    loading: {
      "props": {
        "qhf8xq": _u1u1p7mj.default,
        "Bhzewxz": _ufibavv.default,
        "oyh7mz": _u1jmni.default,
        "Bz10aip": _uevjesf.default,
        "mc9l5x": _ukchsq.default,
        "Brf1p80": _ut7q8es.default,
        "Bt984gj": _ubyhozq.default
      },
      "alias": `${prefix}__loading`
    },
    error: {
      "props": {
        "qhf8xq": _u1u1p7mj.default,
        "Bhzewxz": _ufibavv.default,
        "oyh7mz": _u1jmni.default,
        "Bz10aip": _uevjesf.default,
        "mc9l5x": _ukchsq.default,
        "Brf1p80": _ut7q8es.default,
        "Bt984gj": _ubyhozq.default,
        "a9b677": _u1it5fv.default,
        "Bqenvij": _u1b894eo.default,
        "ayd6f0": _uhiotwf.default,
        "B4j52fo": _u1cpcpb.default,
        "Bekrc4i": _u10qn3go.default,
        "Bn0qgzm": _uflkg.default,
        "ibv6hh": _uvn5txf.default,
        "icvyot": _us9j13g.default,
        "vrafjx": _umt7kug.default,
        "oivjwe": _uh3lwu.default,
        "wvpqe5": _ur.default,
        "g2u3we": _u1n7c7ic.default,
        "h3c5rm": _ujkvyav.default,
        "B9xav0g": _u182xfmz.default,
        "zhjwy3": _uqbl7tl.default,
        "Btl43ni": _urnexrx.default,
        "B7oj6ja": _u12hcvx.default,
        "Bbmb7ep": _u1gja64x.default,
        "Beyfa6y": _u162mx8s.default
      },
      "alias": `${prefix}__error`
    },
    errorIcon: {
      "props": {
        "sj55zd": _u7iz6qb.default,
        "Be2twd7": _ugj19bf.default
      },
      "alias": `${prefix}__error__icon`
    },
    hideImage: {
      "props": {
        "abs64n": _usfh9mg.default
      },
      "alias": `${prefix}__hide`
    }
  };
});
const useImageViewerImage = () => {
  const {
    initTransform,
    setMatrixByPoints,
    scaleByOrigin,
    setDialogVisible
  } = (0, _ImageViewerContext.useImageViewerContext)();
  const [isImageVisible, setIsImageVisible] = React.useState(false);
  const [isPressing, setIsPressing] = React.useState(false);
  const [prevSrc, setPrevSrc] = React.useState('');

  /**
   * Translation
   *
   * - 按住拖拽（跨端同构，单点，onPointerDown + Move + Up）
   */
  const singleTouchTranslationHandler = (0, _usePointerHandler.usePointerHandler)({
    down: () => {
      setIsPressing(true);
    },
    move: (ev, pointerIds) => {
      const lastEvent = pointerIds.get(ev.pointerId);
      if (pointerIds.size === 1 && lastEvent) {
        // 单点触控
        const {
          clientX,
          clientY
        } = lastEvent;
        setMatrixByPoints(point => {
          const translation = _glMatrix.mat2d.fromTranslation(_glMatrix.mat2d.create(), [ev.clientX - clientX, ev.clientY - clientY]);
          return _glMatrix.vec2.transformMat2d(_glMatrix.vec2.create(), point, translation);
        });
      }
    },
    up: () => {
      setIsPressing(false);
    },
    tap: event => {
      setDialogVisible(false, {
        visible: false,
        event,
        action: _useImageViewerVisible.ImageViewerCloseAction.Image
      });
    }
  });

  /**
   * Translation
   *
   * - 触控板滚动（onWheel）
   * - 鼠标滚轮（onWheel）
   */
  React.useEffect(() => {
    const handleWheel = event => {
      const {
        metaKey,
        ctrlKey
      } = event;
      if (metaKey || ctrlKey) {
        return;
      }
      setMatrixByPoints(point => {
        const translation = _glMatrix.mat2d.fromTranslation(_glMatrix.mat2d.create(), [-event.deltaX, -event.deltaY]);
        return _glMatrix.vec2.transformMat2d(_glMatrix.vec2.create(), point, translation);
      });
    };
    document.addEventListener('wheel', handleWheel);
    return () => {
      document.removeEventListener('wheel', handleWheel);
    };
  });

  /**
   * Scaling
   * - 触控板 pinch（onWheel ctrl）
   * - 触控板 + meta（onWheel + meta）
   * - 鼠标滚轮 + ctrl（onWheel ctrl）
   */
  React.useEffect(() => {
    const handleWheel = event => {
      event.preventDefault();
      const {
        deltaY,
        ctrlKey,
        metaKey
      } = event;
      if (metaKey || ctrlKey) {
        /**
         * 1. ctrl 或者 cmd 加 wheel 进行缩放
         * 2. 触控板 pinch 缩放，浏览器默认会加上 ctrlKey true
         */
        scaleByOrigin({
          origin: _glMatrix.vec2.fromValues(event.clientX, event.clientY),
          getNextScale: currentScale => currentScale * (_constants.CURRENT_SCALE_OF_ORIGINAL_SIZE - deltaY * _constants.DELTA_Y_SCALE)
        });
      }
    };
    document.addEventListener('wheel', handleWheel, {
      passive: false
    });
    return () => {
      document.removeEventListener('wheel', handleWheel);
    };
  }, [scaleByOrigin]);
  const mobileScalingHandler = (0, _useImageViewerMobileScaling.useImageViewerMobileScaling)();
  const imageProps = {
    onLoad: () => {
      initTransform({
        fitToContainer: true
      });
      setIsImageVisible(true);
    },
    onPointerDown: (0, _reactUtilities.useEventCallback)((0, _function.composeEventHandlers)(singleTouchTranslationHandler, mobileScalingHandler))
  };
  return {
    slotsProps: {
      image: imageProps
    },
    state: {
      isImageVisible,
      setIsImageVisible,
      prevSrc,
      setPrevSrc,
      isPressing
    }
  };
};
const ImageViewerImage = exports.ImageViewerImage = (0, _reactUtilities.forwardRef)((props, ref) => {
  var _image$alt;
  const {
    errorFallback
  } = props;
  const {
    rootCssScope,
    imageRef,
    image,
    isImageContained
  } = (0, _ImageViewerContext.useImageViewerContext)();
  const {
    state: {
      status
    },
    slotsProps: useImageSlotsProps
  } = (0, _Image.useImage_unstable)(image, (0, _reactUtilities.useMergedRefs)(ref, imageRef));
  const {
    slotsProps: useImageViewerImageSlotsProps,
    state: useImageViewerImageState
  } = useImageViewerImage();
  const {
    isImageVisible,
    setIsImageVisible,
    prevSrc,
    setPrevSrc,
    isPressing
  } = useImageViewerImageState;
  const styles = useImageViewerImageStylexStyles({
    prefix: rootCssScope('__image')
  });
  if (image.src !== prevSrc) {
    setIsImageVisible(false);
    setPrevSrc(image.src);
  }

  // 获取光标形状
  const cursor = (() => {
    if (isImageContained) {
      return 'zoom-out';
    }
    if (isPressing) {
      return 'grabbing';
    }
    return 'grab';
  })();
  return (0, _react.jsx)(React.Fragment, null, (0, _react.jsx)("img", (0, _extends2.default)({}, useImageSlotsProps.root, useImageViewerImageSlotsProps.image, {
    sx: [useImageSlotsProps.root.sx, styles.image, status === 'error' && styles.imageError, !isImageVisible && styles.hideImage],
    style: {
      cursor,
      ...useImageSlotsProps.root.style
    },
    alt: (_image$alt = image === null || image === void 0 ? void 0 : image.alt) !== null && _image$alt !== void 0 ? _image$alt : 'preview',
    onLoad: (0, _function.composeEventHandlers)(useImageSlotsProps.root.onLoad, useImageViewerImageSlotsProps.image.onLoad)
  })), status === 'error' && (0, _slot.resolveSlot)(errorFallback, {
    defaultNode: (0, _react.jsx)("div", {
      sx: [styles.error]
    }, (0, _react.jsx)(_LoadfailFilled.default, {
      className: (0, _stylex.default)(styles.errorIcon)
    }))
  }), status === 'loading' && (0, _react.jsx)("div", {
    sx: [styles.loading]
  }, (0, _react.jsx)(_loading.Loading, {
    size: "large"
  })));
});
