// @universe-design/react/lib/image-viewer/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ImageViewer", {
  enumerable: true,
  get: function () {
    return _ImageViewer.ImageViewer;
  }
});
Object.defineProperty(exports, "ImageViewerCloseAction", {
  enumerable: true,
  get: function () {
    return _useImageViewerVisible.ImageViewerCloseAction;
  }
});
Object.defineProperty(exports, "useImageViewerContext", {
  enumerable: true,
  get: function () {
    return _ImageViewerContext.useImageViewerContextExternal;
  }
});
Object.defineProperty(exports, "useImageViewerToolbarDownload", {
  enumerable: true,
  get: function () {
    return _Download.useImageViewerToolbarDownload;
  }
});
Object.defineProperty(exports, "useImageViewerToolbarPagination", {
  enumerable: true,
  get: function () {
    return _Pagination.useImageViewerToolbarPagination;
  }
});
Object.defineProperty(exports, "useImageViewerToolbarRotation", {
  enumerable: true,
  get: function () {
    return _Rotation.useImageViewerToolbarRotation;
  }
});
Object.defineProperty(exports, "useImageViewerToolbarScaling", {
  enumerable: true,
  get: function () {
    return _Scaling.useImageViewerToolbarScaling;
  }
});
var _ImageViewerContext = require("./ImageViewerContext");
var _ImageViewer = require("./ImageViewer");
var _Download = require("./components/Download");
var _Pagination = require("./components/Pagination");
var _Rotation = require("./components/Rotation");
var _Scaling = require("./components/Scaling");
var _useImageViewerVisible = require("./useImageViewerVisible");
