// @universe-design/react/lib/image-viewer/useImageViewerTransform.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useImageViewerTransform = void 0;
var React = _interopRequireWildcard(require("react"));
var _glMatrix = require("gl-matrix");
var _reactUtilities = require("@universe-design/react-utilities");
var _transform = require("./transform");
var _constants = require("./constants");
var _assert = require("../util/assert");
/**
 * ================================
 *
 * useImageViewerTransform
 *
 * ================================
 */

/**
 * 或者图片的原始大小
 */
const getImageOriginSize = el => {
  var _el$naturalWidth, _el$naturalHeight;
  return {
    width: (_el$naturalWidth = el === null || el === void 0 ? void 0 : el.naturalWidth) !== null && _el$naturalWidth !== void 0 ? _el$naturalWidth : 0,
    height: (_el$naturalHeight = el === null || el === void 0 ? void 0 : el.naturalHeight) !== null && _el$naturalHeight !== void 0 ? _el$naturalHeight : 0
  };
};

/**
 * 与图片变换相关的逻辑
 */
const useImageViewerTransform = props => {
  const {
    minScale = _constants.MIN_SCALE_NUM,
    maxScale = _constants.MAX_SCALE_NUM
  } = props;
  const imageRef = React.useRef(null);
  const [matrix, _setMatrix, matrixRef] = (0, _reactUtilities.useLatestState)(_glMatrix.mat2d.create());
  const [containerSize, setContainerSize] = React.useState({
    width: 0,
    height: 0
  });
  const containerRef = (0, _reactUtilities.useResizeObserver)(entry => {
    const imageOriginalRect = getImageOriginalRect();
    const imageOriginalPoints = (0, _transform.rectToPoints)(imageOriginalRect);
    const nextPoints = imageOriginalPoints.map(p => _glMatrix.vec2.transformMat2d(_glMatrix.vec2.create(), p, matrixRef.current));
    const containerRect = getContainerRect(entry);
    const clampedMat = getClampedMatrix({
      currentPoints: imageOriginalPoints,
      nextPoints,
      containerRect
    });
    clampedMat && setMatrix(clampedMat);
    setContainerSize((0, _reactUtilities.getSize)(entry.target, entry));
  });

  /**
   * 获取夹过的变换矩阵
   */
  const getClampedMatrix = (0, _reactUtilities.useEventCallback)(_ref => {
    let {
      currentPoints,
      nextPoints,
      containerRect
    } = _ref;
    const clampedPoints = (0, _transform.clampPoints)(nextPoints, {
      containerRect
    });
    return (0, _transform.getTransformBetweenTwoRects)(currentPoints, clampedPoints);
  });

  /**
   * Matrix
   */
  const getMatrix = (0, _reactUtilities.useEventCallback)(() => matrixRef.current);
  const setMatrix = (0, _reactUtilities.useEventCallback)(mat => {
    const image = imageRef.current;
    if (image) {
      // 我们使用原生 dom 操作跳过 react 直接调整图片变换
      image.style.transform = `matrix(${mat.join(', ')})`;
    }
    _setMatrix(mat);
  });

  /**
   * Image
   */
  const [imageOriginalSize, setImageOriginalSize] = React.useState({
    width: 0,
    height: 0
  });
  const getImageOriginalRect = (0, _reactUtilities.useEventCallback)(() => {
    const {
      width,
      height
    } = getImageOriginSize(imageRef.current);
    return new DOMRect(0, 0, width, height);
  });
  const setMatrixByPoints = (0, _reactUtilities.useEventCallback)(pointsOrGetNextPoint => {
    const originalPoints = (0, _transform.rectToPoints)(getImageOriginalRect());
    const currentMat = getMatrix();
    const nextPoints = (0, _assert.isFunction)(pointsOrGetNextPoint) ? originalPoints.map(point => pointsOrGetNextPoint(_glMatrix.vec2.transformMat2d(_glMatrix.vec2.create(), point, currentMat))) : pointsOrGetNextPoint;
    const containerRect = getContainerRect();
    const clampedMat = getClampedMatrix({
      currentPoints: originalPoints,
      nextPoints,
      containerRect
    });
    setMatrix(clampedMat);
  });

  /**
   * Container
   */
  const getContainerRect = (0, _reactUtilities.useEventCallback)(entry => {
    const container = containerRef.current;
    if (!container) {
      return new DOMRect();
    }
    const containerSize = (0, _reactUtilities.getSize)(container, entry);
    const containerRect = new DOMRect(0, 0, containerSize.width, containerSize.height);
    return containerRect;
  });
  const initTransform = (0, _reactUtilities.useEventCallback)(function () {
    let options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    const {
      fitToContainer = true
    } = options;
    const imageOriginalRect = getImageOriginalRect();
    const container = containerRef.current;
    if (!imageOriginalRect || !container) {
      return;
    }
    // 用 layout size 而不是 bounding rect 可以消除动画带来的影响
    const {
      clientWidth: containerWidth,
      clientHeight: containerHeight
    } = container;
    const nextImageRect = (() => {
      let width = imageOriginalRect.width;
      let height = imageOriginalRect.height;
      if (fitToContainer) {
        // 算出 x y 完全卡到 boundary 时的缩放比，取小的那个可以保证是 contain 的效果
        const scale = Math.min(Math.max(Math.min((containerWidth - _constants.FIT_TO_CONTAINER_BOUNDARY_PADDING * 2) / imageOriginalRect.width, (containerHeight - _constants.FIT_TO_CONTAINER_BOUNDARY_PADDING * 2) / imageOriginalRect.height, 1), minScale), maxScale);
        width = imageOriginalRect.width * scale;
        height = imageOriginalRect.height * scale;
      }
      // TODO 这里其实有问题，如果 container x 和 y 不在 0 就不对了
      return new DOMRect((containerWidth - width) / 2, (containerHeight - height) / 2, width, height);
    })();
    setImageOriginalSize(imageOriginalRect);
    setMatrix((0, _transform.getTransformBetweenTwoRects)((0, _transform.rectToPoints)(imageOriginalRect), (0, _transform.rectToPoints)(nextImageRect)));
  });
  const getScale = (0, _reactUtilities.useEventCallback)(() => (0, _transform.getMatScale)(matrix));

  /**
   * 按中心缩放
   */
  const scaleByOrigin = (0, _reactUtilities.useEventCallback)(_ref2 => {
    let {
      origin: originParam,
      getNextScale
    } = _ref2;
    const origin = (() => {
      if (originParam) {
        return originParam;
      }
      const containerRect = getContainerRect();
      return _glMatrix.vec2.fromValues((containerRect.x + containerRect.width) / 2, (containerRect.y + containerRect.height) / 2);
    })();
    const currentScale = getScale();
    const nextScale = Math.max(minScale, Math.min(getNextScale(currentScale), maxScale)); // Limit the scale

    if (nextScale === currentScale) {
      return;
    }
    const scaleMat = _glMatrix.mat2d.fromScaling(_glMatrix.mat2d.create(), [nextScale / currentScale, nextScale / currentScale]);
    setMatrixByPoints(absolutePoint => {
      const point = _glMatrix.vec2.subtract(_glMatrix.vec2.create(), absolutePoint, origin); // 基于 origin 的坐标
      let nextPoint = _glMatrix.vec2.create();
      nextPoint = _glMatrix.vec2.transformMat2d(_glMatrix.vec2.create(), point, scaleMat);
      nextPoint = _glMatrix.vec2.add(_glMatrix.vec2.create(), nextPoint, origin);
      return nextPoint;
    });
  });

  /**
   * 这里用了一些场外信息，我们的 image 只会做 scaleX 等于 scaleY 的等比缩放
   */
  const scale = (0, _transform.getMatScale)(matrix);
  const rotationRadians = (0, _transform.getMatRotationRadians)(matrix);

  /**
   * 判断图片尺寸是否大于容器尺寸
   *  */
  const isImageContained = React.useMemo(() => {
    if (typeof DOMRect === 'undefined') {
      return false;
    }
    const imageOriginalRect = new DOMRect(0, 0, imageOriginalSize.width, imageOriginalSize.height);
    const originalPoints = [...(0, _transform.rectToPoints)(imageOriginalRect), _glMatrix.vec2.fromValues(imageOriginalRect.x + imageOriginalRect.width, imageOriginalRect.y + imageOriginalRect.height)];
    const transformedPoints = [_glMatrix.vec2.create(), _glMatrix.vec2.create(), _glMatrix.vec2.create()];
    for (const [i, transformedPoint] of transformedPoints.entries()) {
      _glMatrix.vec2.transformMat2d(transformedPoint, originalPoints[i], matrix);
    }
    return (0, _transform.isPointsInRectangleRegion)(new DOMRect(0, 0, containerSize.width, containerSize.height), transformedPoints);
  }, [containerSize.height, containerSize.width, imageOriginalSize.height, imageOriginalSize.width, matrix]);
  return {
    state: {
      getMatrix,
      setMatrix,
      setMatrixByPoints,
      imageRef,
      containerRef,
      getContainerRect,
      initTransform,
      getScale,
      scale,
      rotationRadians,
      scaleByOrigin,
      minScale,
      maxScale,
      /**
       * 判断图片区域是否超出容器区域
       */
      isImageContained
    }
  };
};
exports.useImageViewerTransform = useImageViewerTransform;
