// @universe-design/react/lib/image-viewer/useImageViewerVisible.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useImageViewerVisible = exports.ImageViewerCloseAction = void 0;
var _reactUtilities = require("@universe-design/react-utilities");
var _util = require("../util");
var _hooks = require("../hooks");
/**
 * ================================
 *
 * useImageViewerVisible
 *
 * ================================
 */

/**
 * 与图片查看器弹层控制相关的数据逻辑
 */
const useImageViewerVisible = props => {
  var _visible$defaultValue;
  const {
    visible
  } = props;
  const [dialogVisible, setDialogVisible] = (0, _reactUtilities.useControlledState)((_visible$defaultValue = visible === null || visible === void 0 ? void 0 : visible.defaultValue) !== null && _visible$defaultValue !== void 0 ? _visible$defaultValue : false, visible === null || visible === void 0 ? void 0 : visible.value);
  const setDialogVisibleWithMeta = (0, _hooks.useEventCallback)((value, meta) => {
    setDialogVisible(value);
    if (!visible) {
      return;
    }
    const {
      onChange
    } = visible;
    (0, _util.isFunction)(onChange) && onChange(value, meta);
  });
  return {
    state: {
      dialogVisible,
      setDialogVisible: setDialogVisibleWithMeta
    }
  };
};

/**
 * 图片查看器可见性相关属性
 */

/** 关闭图片查看器的参数 */
exports.useImageViewerVisible = useImageViewerVisible;
/**
 * 图片查看器关闭动作
 */
let ImageViewerCloseAction = exports.ImageViewerCloseAction = /*#__PURE__*/function (ImageViewerCloseAction) {
  ImageViewerCloseAction["Mask"] = "mask";
  ImageViewerCloseAction["CloseButton"] = "closeButton";
  ImageViewerCloseAction["Esc"] = "esc";
  ImageViewerCloseAction["Image"] = "image";
  return ImageViewerCloseAction;
}({});
