// @universe-design/react/es/shared/components/trigger/modifiers/useTriggerArrow.js
import clsx from 'clsx';
import React, { useMemo } from 'react';
import { isFunction, isNonNullable } from '../../../../util';
export const useTriggerArrow = config => {
  const {
    childSize,
    popupSize,
    arrow,
    popup: popupProp,
    prefixCls
  } = config;
  const className = `${prefixCls}__arrow`;
  const showArrow = isNonNullable(arrow);
  const {
    // 这个 size 是猪整个方块的对角线一半，但是因为 arrow 有 border 所以整体会内陷 1px，实际暴露出去的是 6px arrow size
    size: arrowSize = 8,
    // 默认用 ud 的中号 border-radius
    safeMarginX: arrowSafeMarginX = 8,
    safeMarginY: arrowSafeMarginY = 8,
    content
  } = arrow !== null && arrow !== void 0 ? arrow : {};

  // thunk for better code style
  const placement = (() => {
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    const normalizeAngle = a => (a < 0 ? Math.PI * 2 + a : a) / Math.PI * 180;
    // 计算 trigger 重心和 popup 重心之间的夹角，由于浏览器坐标系 y 轴是反的，所以 y 参数计算的时候要反向
    const trigger2PopupAngle = normalizeAngle(Math.atan2((childSize.y + childSize.height / 2 - (popupSize.y + popupSize.height / 2)) * -1, childSize.x + childSize.width / 2 - (popupSize.x + popupSize.width / 2)));
    const angles = {
      tr: normalizeAngle(Math.atan2(popupSize.height, popupSize.width)),
      tl: normalizeAngle(Math.atan2(popupSize.height, -popupSize.width)),
      bl: normalizeAngle(Math.atan2(-popupSize.height, -popupSize.width)),
      br: normalizeAngle(Math.atan2(-popupSize.height, popupSize.width))
    };
    if (trigger2PopupAngle >= angles.tr && trigger2PopupAngle < angles.tl) {
      return 'top';
    } else if (trigger2PopupAngle >= angles.tl && trigger2PopupAngle < angles.bl) {
      return 'left';
    } else if (trigger2PopupAngle >= angles.bl && trigger2PopupAngle < angles.br) {
      return 'bottom';
    } else {
      return 'right';
    }
  })();

  // console.log(childSize, popupSize);

  // thunk for better code style

  const contentNode = (() => {
    if (isFunction(content)) {
      return content({
        placement
      });
    }
    if (isNonNullable(content)) {
      return /*#__PURE__*/React.createElement("div", {
        className: `${className}__content`,
        style: {
          width: arrowSize * Math.SQRT2,
          height: arrowSize * Math.SQRT2
        }
      });
    }
    return content;
  })();
  const arrowNode = showArrow ? /*#__PURE__*/React.createElement("div", {
    className: clsx(className, `${className}-${placement}`),
    style: (() => {
      const style = {};

      // 如果是在上下出现 arrow
      if (['bottom', 'top'].includes(placement)) {
        style.left = childSize.left + childSize.width / 2 - popupSize.left;
        style.width = arrowSize * 2;
        style.height = arrowSize;
        if (style.left - arrowSize - arrowSafeMarginX < 0) {
          style.left = arrowSize + arrowSafeMarginX;
          // 设计允许三角在极端 case 下不指向 trigger 元素中心，所以把这个 size reset 逻辑注释掉
          // TODO: 目前还有一个 1-3 2-4 象限的极端 case 由于先判断 vertical 方向，所以永远都会走 vertical 的 branch
          // style.width = 0;
          // style.height = 0;
        } else if (style.left + arrowSize + arrowSafeMarginX > popupSize.width) {
          style.left = popupSize.width - arrowSize - arrowSafeMarginX;
          // style.width = 0;
          // style.height = 0;
        }
      } else if (['left', 'right'].includes(placement)) {
        // 如果是在左右出现 arrow
        style.top = childSize.top + childSize.height / 2 - popupSize.top;
        style.width = arrowSize;
        style.height = arrowSize * 2;
        if (style.top - arrowSize - arrowSafeMarginY < 0) {
          style.top = arrowSize + arrowSafeMarginY;
          // style.width = 0;
          // style.height = 0;
        } else if (style.top + arrowSize + arrowSafeMarginY > popupSize.height) {
          style.top = popupSize.height - arrowSize - arrowSafeMarginY;
          // style.width = 0;
          // style.height = 0;
        }
      }

      // 其他情况下不显示箭头
      return style;
    })()
  }, contentNode) : null;
  const origPopupNode = useMemo(() => isFunction(popupProp) ? popupProp() : popupProp, [popupProp]);
  return /*#__PURE__*/React.createElement(React.Fragment, null, origPopupNode, arrowNode);
};
