// @universe-design/react/es/shared/components/trigger/modifiers/useTriggerObserver.js
import { useState, useCallback } from 'react';
import { unstable_batchedUpdates } from 'react-dom';
import { alignElement } from 'dom-align';
import { isElement } from '../@popperjs/core/lib/dom-utils/instanceOf';
import listScrollParents from '../@popperjs/core/lib/dom-utils/listScrollParents';
import { useIsomorphicLayoutEffect as useLayoutEffect } from '../../../../hooks';
import isEqual from 'lodash/isEqual';
const defaultSize = {
  width: 0,
  height: 0,
  top: 0,
  left: 0,
  right: 0,
  bottom: 0,
  x: 0,
  y: 0
};
const getPositionSize = rect => {
  // Safari 浏览器下 rect 尺寸小数点后几位会不稳定
  const left = Math.floor(rect.left);
  const top = Math.floor(rect.top);
  const width = Math.floor(rect.width);
  const height = Math.floor(rect.height);
  return {
    left,
    top,
    x: left,
    y: top,
    width,
    height,
    right: width + left,
    bottom: height + top
  };
};

/**
 * dom-align 用于获取可视区域的函数
 */
const getVisibleRectForElement = alignElement.__getVisibleRectForElement;
export const useTriggerObserver = config => {
  const {
    popupAlign,
    triggerRef,
    readyObserver,
    enablePopupOverflowAuto = false,
    enableforcePopupAlign = false
  } = config;
  const [popupSize, setPopupSize] = useState(defaultSize);
  const [childSize, setChildSize] = useState(defaultSize);
  const [viewPointRect, setViewPointRect] = useState(null);
  const updateRect = useCallback(() => {
    var _triggerRef$current, _triggerRef$current2;
    const child = (_triggerRef$current = triggerRef.current) === null || _triggerRef$current === void 0 ? void 0 : _triggerRef$current.getRootDomNode();
    const popup = (_triggerRef$current2 = triggerRef.current) === null || _triggerRef$current2 === void 0 ? void 0 : _triggerRef$current2.getPopupDomNode();
    if (!child || !popup) {
      return;
    }
    unstable_batchedUpdates(() => {
      const newPopupSize = getPositionSize(popup.getBoundingClientRect());
      const newChildSize = getPositionSize(child.getBoundingClientRect());
      if (!isEqual(newPopupSize, popupSize)) {
        setPopupSize(newPopupSize);
      }
      if (!isEqual(newChildSize, childSize)) {
        setChildSize(newChildSize);
      }
      if (enablePopupOverflowAuto) {
        var _popupAlign$overflow;
        // FIXME 现在这里逻辑还有点问题，如果 visible rect 大小变了其实我们是无法得到监听的
        const alwaysByViewport = Boolean((_popupAlign$overflow = popupAlign.overflow) === null || _popupAlign$overflow === void 0 ? void 0 : _popupAlign$overflow.alwaysByViewport);
        const newViewPointRect = getVisibleRectForElement(popup, alwaysByViewport);
        if (!isEqual(newViewPointRect, viewPointRect)) {
          setViewPointRect(newViewPointRect);
        }
      }
    });
  }, [popupAlign, triggerRef, popupSize, childSize, viewPointRect, enablePopupOverflowAuto]);
  useLayoutEffect(() => {
    var _triggerRef$current3, _triggerRef$current4;
    if (!readyObserver) {
      return;
    }
    const passive = {
      passive: true
    };
    const child = (_triggerRef$current3 = triggerRef.current) === null || _triggerRef$current3 === void 0 ? void 0 : _triggerRef$current3.getRootDomNode();
    const popup = (_triggerRef$current4 = triggerRef.current) === null || _triggerRef$current4 === void 0 ? void 0 : _triggerRef$current4.getPopupDomNode();
    if (!child || !popup) {
      return;
    }
    updateRect();
    const forceAlign = () => {
      var _triggerRef$current5;
      (_triggerRef$current5 = triggerRef.current) === null || _triggerRef$current5 === void 0 ? void 0 : _triggerRef$current5.forcePopupAlign();
    };
    let scrollParents = [];
    if (enableforcePopupAlign) {
      scrollParents = [...new Set([...(isElement(child) ? listScrollParents(child) : []), ...(isElement(popup) ? listScrollParents(popup) : [])])];
      scrollParents.forEach(scrollParent => {
        scrollParent.addEventListener('scroll', forceAlign, passive);
      });
      window.addEventListener('resize', forceAlign, passive);
    }

    // 观察器的配置（需要观察什么变动）

    // 创建一个观察器实例并传入回调函数
    const mutationObserver = new MutationObserver(mutationsList => {
      // 如果定位发生变化，重新计算位置
      const observeStyleKey = ['left', 'top', 'bottm', 'right'];
      const oldValue = new Map();

      // Use traditional 'for loops' for IE 11
      for (const mutation of mutationsList) {
        if (mutation.type === 'attributes' && mutation.attributeName === 'style') {
          const trim = str => str.trim();
          if (typeof mutation.oldValue === 'string') {
            const values = mutation.oldValue.split(';').map(trim).filter(Boolean).map(e => e.split(':').map(trim));
            values.forEach(_ref => {
              let [key, value] = _ref;
              if (!observeStyleKey.includes(key)) {
                return;
              }
              if (oldValue.has(key)) {
                return;
              }
              oldValue.set(key, value);
            });
          }
        }
      }
      let isChanged = false;
      oldValue.forEach((value, key) => {
        if (value !== popup.style[key]) {
          isChanged = true;
        }
      });
      if (isChanged) {
        updateRect();
      }
    });
    mutationObserver.observe(popup, {
      attributes: true,
      childList: false,
      subtree: false,
      attributeOldValue: true,
      attributeFilter: ['style']
    });
    return () => {
      if (enableforcePopupAlign) {
        scrollParents.forEach(scrollParent => {
          scrollParent.removeEventListener('scroll', forceAlign, !passive.passive);
        });
        window.removeEventListener('resize', forceAlign, !passive.passive);
      }
      mutationObserver.disconnect();
    };
  }, [readyObserver, updateRect, triggerRef, enableforcePopupAlign]);
  return [childSize, popupSize, viewPointRect];
};
