// @byted-service/fetch-core/lib/helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.encodeUnicodeText = exports.parseAddr = exports.proxyAgent = void 0;
const tslib_1 = require("tslib");
const url_1 = require("url");
const http_proxy_agent_1 = tslib_1.__importDefault(require("http-proxy-agent"));
const https_proxy_agent_1 = tslib_1.__importDefault(require("https-proxy-agent"));
const enableProxy = process.env.ENABLE_FETCH_PROXY === 'true';
const proxySym = Symbol.for('proxy');
const proxyAgents = {};
const proxyAgent = (url, options) => {
    if (options.enableProxy !== true && !enableProxy) {
        return;
    }
    let proxy = options.proxy;
    if (!proxy) {
        switch (url.protocol) {
            case 'http:':
                proxy = process.env.HTTP_PROXY || process.env.http_proxy;
                break;
            case 'https:':
                proxy = process.env.HTTPS_PROXY || process.env.https_proxy;
                break;
        }
    }
    if (!proxy) {
        return;
    }
    let proxyAgent = proxyAgents[`${url.protocol}-${proxy}`];
    if (!proxyAgent) {
        switch (url.protocol) {
            case 'http:':
                proxyAgent = (0, http_proxy_agent_1.default)(proxy);
                break;
            case 'https:':
                proxyAgent = (0, https_proxy_agent_1.default)(proxy);
                break;
        }
        if (proxyAgent) {
            proxyAgent[proxySym] = new url_1.URL(proxy);
            proxyAgents[`${url.protocol}-${proxy}`] = proxyAgent;
        }
    }
    return proxyAgent;
};
exports.proxyAgent = proxyAgent;
/**
 * 只兼容以下集中格式
 * - /path/to/xx.sock
 * - 127.0.0.1:3306
 * - [::]:3306
 */
const parseAddr = (addr) => {
    let parts;
    switch (true) {
        case addr.startsWith('/'):
            return { host: '', port: '', socketPath: addr };
        case addr.startsWith('['):
            parts = addr.split(']');
            return {
                host: parts[0].slice(1),
                port: parts[1].slice(1),
                socketPath: '',
            };
        default:
            parts = addr.split(':');
            return { host: parts[0], port: parts[1], socketPath: '' };
    }
};
exports.parseAddr = parseAddr;
const encodeUnicodeText = (text) => {
    return Array.from(text)
        .map((char) => {
        if (char.charCodeAt(0) > 127) {
            return '\\u' + char.charCodeAt(0).toString(16).padStart(4, '0');
        }
        return char;
    })
        .join('');
};
exports.encodeUnicodeText = encodeUnicodeText;
