// @byted-service/fetch-core/lib/metrics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sendMetrics = exports.metricsNameSpace = exports.metricsLabel = exports.metricsNames = void 0;
const net_1 = require("net");
const helper_1 = require("./helper");
exports.metricsNames = {
    rpcSuccessThroughput: (psm) => `service.thrift.${psm}.call.success.throughput`,
    rpcErrorThroughput: (psm) => `service.thrift.${psm}.call.error.throughput`,
    rpcSuccessLatency: (psm) => `service.thrift.${psm}.call.success.latency.us`,
    rpcErrorLatency: (psm) => `service.thrift.${psm}.call.error.latency.us`,
};
exports.metricsLabel = {
    success: 'success',
    rpcError: 'net_err',
    bizError: 'business_err',
};
exports.metricsNameSpace = 'toutiao';
function sendMetrics(ctx) {
    if (ctx.from === '-') {
        return;
    }
    const tags = Object.assign({ 
        // basic
        method: ctx.toMethod, from_method: ctx.fromMethod, env: ctx.env || 'prod', stress_tag: ctx.stressTag || '-', 
        // from / to
        from: ctx.from, from_cluster: ctx.fromCluster, from_idc: ctx.fromIdc, to: ctx.to, to_cluster: ctx.toCluster, to_idc: ctx.toIdc || '-', 
        // call
        status: ctx.status, label: exports.metricsLabel.success, retry: ctx.times, 
        // ip
        mesh: ctx.mesh === '1' ? '1' : '0', ipv6: ctx.rip && (0, net_1.isIPv6)((0, helper_1.parseAddr)(ctx.rip).host) ? '1' : '0', 
        // rip: host || rip,
        // ttlogagent
        remote_protocol: ctx.remoteProtocol }, ctx.metricsTags);
    if (ctx.status >= 200 && ctx.status < 400) {
        ctx.metrics.emitTimer(exports.metricsNames.rpcSuccessLatency(ctx.from), ctx.timer.cost('μs'), exports.metricsNameSpace, tags);
        ctx.metrics.emitCounter(exports.metricsNames.rpcSuccessThroughput(ctx.from), 1, exports.metricsNameSpace, tags);
    }
    else {
        tags.label = exports.metricsLabel.bizError;
        ctx.metrics.emitTimer(exports.metricsNames.rpcErrorLatency(ctx.from), ctx.timer.cost('μs'), exports.metricsNameSpace, tags);
        ctx.metrics.emitCounter(exports.metricsNames.rpcErrorThroughput(ctx.from), 1, exports.metricsNameSpace, tags);
    }
}
exports.sendMetrics = sendMetrics;
