// @byted-service/logger/lib/core_logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CoreLogger = void 0;
const util_1 = require("util");
const flusher_1 = require("./flusher");
const consts_1 = require("./consts");
const helper_1 = require("./helper");
const flusher = new flusher_1.Flusher();
class CoreLogger {
    constructor(options = {}) {
        this.levels = consts_1.levels;
        this.streams = [];
        this.level = options.level || consts_1.defaultLogLevel;
        this.defaultLevel = options.defaultLevel || consts_1.defaultLevel;
        this.setFormatter(options.formatter || { format: (msg) => msg });
        this.flushInterval = options.flushInterval || consts_1.defaultFlushInterval;
        const streams = options.streams || [];
        for (let i = 0; i < streams.length; i++) {
            this.addStream(streams[i].level || this.level, streams[i].stream);
        }
        flusher.register(this.flushInterval, this);
    }
    get level() {
        return this._level;
    }
    set level(level) {
        if (this.levels[level] === undefined) {
            throw new Error('unknown level ' + level);
        }
        this._level = level;
        const levels = Object.keys(this.levels);
        for (let i = 0; i < levels.length; i++) {
            this[levels[i]] = this.isLevelEnabled(levels[i])
                ? this.genLog(levels[i])
                : consts_1.noop;
        }
        const fatal = this.fatal.bind(this);
        this.fatal = (...args) => {
            fatal(...args);
            try {
                this.flushSync();
            }
            catch (err) {
                // https://github.com/pinojs/pino/pull/740#discussion_r346788313
            }
        };
    }
    setFormatter(formatter) {
        this.formatter =
            typeof formatter === 'function' ? { format: formatter } : formatter;
    }
    getFormatter() {
        return this.formatter;
    }
    isLevelEnabled(level) {
        return this.levels[level] >= this.levels[this.level];
    }
    log(options = {}) {
        const level = options.level || this.defaultLevel;
        if (!this.isLevelEnabled(level)) {
            return;
        }
        const message = options.message === undefined ? '' : options.message;
        const splat = options.splat || [];
        this.write(level, message, splat, options);
    }
    flush() {
        const streams = this.streams;
        for (let i = 0; i < streams.length; i++) {
            const { stream } = streams[i];
            if (typeof stream.flush === 'function') {
                stream.flush();
            }
        }
    }
    flushSync() {
        const streams = this.streams;
        for (let i = 0; i < streams.length; i++) {
            const { stream } = streams[i];
            if (typeof stream.flushSync === 'function') {
                stream.flushSync();
            }
        }
    }
    clone(options = {}) {
        const logger = Object.create(this);
        logger.level = options.level || this.level;
        logger.defaultLevel = options.defaultLevel || this.defaultLevel;
        logger.setFormatter(options.formatter ||
            (typeof this.formatter.clone === 'function'
                ? this.formatter.clone()
                : this.formatter));
        logger.streams = [];
        logger.flushInterval = options.flushInterval || this.flushInterval;
        const streams = this.streams;
        for (let i = 0; i < streams.length; i++) {
            const { level, stream } = streams[i];
            // level priority:  option level > stream level > logger level
            const cloneLevel = (0, helper_1.formatLogLevel)(options.level, level, logger.level);
            logger.addStream(cloneLevel, typeof stream.clone === 'function' ? stream.clone() : stream);
        }
        if (options.streams) {
            for (let i = 0; i < options.streams.length; i++) {
                logger.addStream(options.streams[i].level || logger.level, options.streams[i].stream);
            }
        }
        return logger;
    }
    addStream(level, stream) {
        this.streams.push({ level, stream });
    }
    deleteStream(level, stream) {
        const streams = this.streams;
        for (let i = 0; i < streams.length; i++) {
            if (streams[i].level === level && streams[i].stream === stream) {
                streams.splice(i, 1);
                break;
            }
        }
    }
    write(level, message, splat, options = {}) {
        options.level = options.level || level;
        options.message = options.message || message;
        options.splat = options.splat || splat;
        const msg = this.formatter.format((0, util_1.format)(message, ...splat), options);
        this.transport(level, msg, options);
    }
    transport(level, message, options = {}) {
        flusher.register(this.flushInterval, this);
        if (message) {
            const streams = this.streams;
            for (let i = 0; i < streams.length; i++) {
                const { level: lv, stream } = streams[i];
                if (this.levels[lv] <= this.levels[level]) {
                    stream.write(message, options);
                }
            }
        }
    }
    genLog(level) {
        return (message, ...splat) => {
            this.write(level, message, splat);
        };
    }
}
exports.CoreLogger = CoreLogger;
