// @byted-service/logger/lib/helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.secMark = exports.toKV = exports.getStackFilePath = exports.formatOptions = exports.formatLogLevel = exports.YYYYMMDDHHmmssSSS = exports.YYYYMMDDHHmmss = void 0;
const tslib_1 = require("tslib");
const path_1 = tslib_1.__importDefault(require("path"));
const dayjs_1 = tslib_1.__importDefault(require("dayjs"));
const lru_cache_1 = tslib_1.__importDefault(require("lru-cache"));
const env_1 = require("@byted-service/env");
const streams_1 = require("./streams");
const consts_1 = require("./consts");
exports.YYYYMMDDHHmmss = (() => {
    let cache = '';
    let lastUpdate = (0, dayjs_1.default)().millisecond(0).toDate().getTime();
    return (now = Date.now()) => {
        if (!cache || now - lastUpdate >= 1000) {
            cache = (0, dayjs_1.default)(now)
                .millisecond(0)
                .tz()
                .format('YYYY-MM-DD HH:mm:ss');
            lastUpdate = now;
        }
        return cache;
    };
})();
const YYYYMMDDHHmmssSSS = (now = Date.now()) => `${(0, exports.YYYYMMDDHHmmss)(now)},${String(now).slice(-3)}`;
exports.YYYYMMDDHHmmssSSS = YYYYMMDDHHmmssSSS;
const formatLogLevel = (...lvs) => {
    for (const level of lvs) {
        if (level && consts_1.levels[level] !== undefined) {
            return level;
        }
    }
    return consts_1.defaultLogLevel;
};
exports.formatLogLevel = formatLogLevel;
const formatOptions = (options = {}) => {
    const { 
    // CoreLogger
    level: customLevel, streams: customStreams, 
    // Logger
    psm: customPsm, prefix: customPrefix, tags: customTags, tagKeys: customTagKeys, callSiteFrameIndex: customCallSiteFrameIndex, enableSecMark: customEnableSecMark, 
    // console
    enableConsoleLog: customEnableConsoleLog, consoleLogLevel: customConsoleLogLevel, 
    // file
    enableFileLog: customEnableFileLog, fileLogLevel: customFileLogLevel, 
    // logagent
    enableAgentLog: customEnableAgentLog, agentLogLevel: customAgentLogLevel, 
    // metrics
    metricsLevel: customMetricsLevel, enableMetrics: customEnableMetrics, } = options;
    const level = (0, exports.formatLogLevel)(customLevel, process.env.LOGGER_OUTPUT_LEVEL, consts_1.defaultLogLevel);
    const streams = customStreams || [];
    // 兼容大多数单测中 mock 后没有设置 getPSM 默认值
    /* istanbul ignore next */
    let psm = (customPsm || (0, env_1.getPSM)() || '').trim();
    if (psm.split('.').length > 3) {
        psm = psm.split('.').slice(0, 3).join('.');
    }
    const prefix = customPrefix || '';
    const tags = customTags;
    const tagKeys = customTagKeys;
    /* istanbul ignore next */
    const callSiteFrameIndex = customCallSiteFrameIndex === undefined ? 1 : customCallSiteFrameIndex;
    const enableSecMark = customEnableSecMark === undefined
        ? consts_1.defaultEnableSecMark
        : customEnableSecMark;
    const consoleLogLevel = (0, exports.formatLogLevel)(customConsoleLogLevel, process.env.LOGGER_LEVEL_CONSOLE, level);
    const fileLogLevel = (0, exports.formatLogLevel)(customFileLogLevel, process.env.LOGGER_LEVEL_FILE, level);
    const agentLogLevel = (0, exports.formatLogLevel)(customAgentLogLevel, process.env.LOGGER_LEVEL_LOGAGENT, level);
    const metricsLevel = (0, exports.formatLogLevel)(customMetricsLevel, process.env.LOGGER_LEVEL_METRICS, consts_1.defaultMetricsLevel);
    // 是否是开发环境，包括本地电脑、开发机、Cloud IDE，只有这三种场景才默认输出控制台日志
    const isDevEnv = (0, env_1.isCloudIDE)() || !(0, env_1.isInTCE)();
    const [enableConsoleLog, enableFileLog, enableAgentLog, enableMetrics] = [
        [
            customEnableConsoleLog,
            process.env.LOGGER_ENABLE_CONSOLE,
            isDevEnv,
        ],
        [customEnableFileLog, process.env.LOGGER_ENABLE_FILE, true],
        [
            customEnableAgentLog,
            process.env.LOGGER_ENABLE_LOGAGENT,
            !isDevEnv,
        ],
        [customEnableMetrics, process.env.LOGGER_ENABLE_METRICS, true],
    ].map(([custom, env, def]) => custom === undefined
        ? env === 'true'
            ? true
            : env === 'false'
                ? false
                : def
        : custom);
    if (enableConsoleLog) {
        streams.push({
            level: consoleLogLevel,
            stream: (0, streams_1.createConsole)(),
        });
    }
    if (enableFileLog) {
        streams.push({
            level: fileLogLevel,
            stream: (0, streams_1.createFileRotate)(Object.assign(Object.assign({}, options), { psm })),
        });
    }
    if (enableAgentLog) {
        streams.push({
            level: agentLogLevel,
            stream: (0, streams_1.createLogAgent)(Object.assign(Object.assign({}, options), { psm })),
        });
    }
    if (enableMetrics) {
        streams.push({
            level: metricsLevel,
            stream: (0, streams_1.createMetrics)(options),
        });
    }
    return {
        level,
        streams,
        psm,
        prefix,
        tags,
        tagKeys,
        callSiteFrameIndex,
        enableSecMark,
    };
};
exports.formatOptions = formatOptions;
const stackTraceCache = new lru_cache_1.default({
    max: Number.isInteger(Number(process.env.LOGGER_STACK_CACHE_SIZE))
        ? /* istanbul ignore next */
            Number(process.env.LOGGER_STACK_CACHE_SIZE)
        : consts_1.defaultStackCacheSize,
});
const prepareStackTrace = (dummyObject, v8StackTrace) => {
    let filepath = '-';
    let lineno = '-';
    const frame = v8StackTrace[Error.stackTraceLimit - 1];
    if (frame) {
        /* istanbul ignore next */
        const filename = frame.getFileName() || '';
        filepath = stackTraceCache.get(filename);
        if (filepath === undefined) {
            filepath = path_1.default.relative(process.cwd(), filename);
            stackTraceCache.set(filename, filepath);
        }
        lineno = frame.getLineNumber();
    }
    return `${filepath}:${lineno}`;
};
function getStackFilePath(callSiteFrameIndex = 0) {
    const oldLimit = Error.stackTraceLimit;
    Error.stackTraceLimit = callSiteFrameIndex;
    const dummyObject = {};
    const v8Handler = Error.prepareStackTrace;
    Error.prepareStackTrace = prepareStackTrace;
    Error.captureStackTrace(dummyObject, getStackFilePath);
    // 避免 v8 懒加载
    // tslint:disable-next-line:no-unused-expression
    dummyObject.stack;
    Error.prepareStackTrace = v8Handler;
    Error.stackTraceLimit = oldLimit;
    return dummyObject.stack;
}
exports.getStackFilePath = getStackFilePath;
function toKV(tags, tagKeys, joiner = ' ', enableSecMark = false) {
    if (!tags) {
        return '';
    }
    if (tagKeys) {
        let str = '';
        if (enableSecMark) {
            for (let i = 0; i < tagKeys.length; i++) {
                str += `{{${tagKeys[i]}=${String(tags[tagKeys[i]])}}}${joiner}`;
            }
        }
        else {
            for (let i = 0; i < tagKeys.length; i++) {
                str += `${tagKeys[i]}=${String(tags[tagKeys[i]])}${joiner}`;
            }
        }
        return str;
    }
    const keys = Object.keys(tags);
    if (keys.length <= 0) {
        return '';
    }
    if (enableSecMark) {
        return `${keys
            .filter((k) => tags.hasOwnProperty(k))
            .map((k) => `{{${k}=${String(tags[k])}}}`)
            .join(joiner)} `;
    }
    else {
        return `${keys
            .filter((k) => tags.hasOwnProperty(k))
            .map((k) => `${k}=${String(tags[k])}`)
            .join(joiner)} `;
    }
}
exports.toKV = toKV;
const secMark = (key, value, enableSecMark = consts_1.defaultEnableSecMark) => enableSecMark ? `{{${key}=${String(value)}}}` : `${key}=${String(value)}`;
exports.secMark = secMark;
