// @universe-design/react/es/date-picker/panels/year-body/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React from 'react';
import PanelBody from '../base/Body';
import { BodyConfig } from '../base/config';
import { useCssScope } from '../../../config-provider/hooks';
import { getYearStart } from '../base/utils';
import { useDateCellInteractiveAreaCls } from '../../cell';
import { isSelectCell } from '../DateCalendar/utils';
import { useDateCalendarUIContext } from '../DateCalendar/DateCalendarUIContext';
function YearBody() {
  const prefixCls = useCssScope('picker')();
  const yearPanelPrefixCls = `${prefixCls}-year-panel`;
  const dateCellInteractiveAreaCls = useDateCellInteractiveAreaCls();
  const {
    shouldCellDisabled,
    pickerValue,
    cellRenderer,
    previewDate,
    selectedDates,
    selectedRange
  } = useDateCalendarUIContext();
  return /*#__PURE__*/React.createElement(PanelBody, _extends({}, BodyConfig.year, {
    className: yearPanelPrefixCls,
    cellClassName: `${yearPanelPrefixCls}-cell`,
    colNum: 4,
    rowNum: 5,
    baseDate: getYearStart(pickerValue.clone()),
    getCellStatus: date => {
      const isRangePreview = previewDate ? previewDate(date) : false;
      const [startDate = null, endDate = null] = selectedRange || [];
      return {
        isDisabled: shouldCellDisabled(date, 'year'),
        isRange: date.isSameOrAfter(startDate, 'year') && date.isSameOrBefore(endDate, 'year'),
        isSelected: isSelectCell({
          mode: 'year',
          selectedDates,
          selectedRange,
          date
        }),
        isRangePreview
      };
    },
    getCellDate: (d, offset) => d.clone().add(offset, 'year'),
    dateRender: d => {
      const node = /*#__PURE__*/React.createElement("div", {
        className: dateCellInteractiveAreaCls
      }, d.format('YYYY'));
      if (typeof cellRenderer === 'function') {
        return cellRenderer(d, {
          viewMode: 'year',
          defaultNode: node
        });
      }
      return node;
    }
  }));
}
export default YearBody;
