// @universe-design/react/src/config-provider/stories/homeless.stories.tsx
import React from 'react';
import type { Meta } from '@storybook/react';
import {
  toast,
  Button,
  ConfigProvider,
  DESKTOP_DEFAULT_CONFIG,
  Avatar,
  useStyleSheetOnce,
} from '@universe-design/react';
import {
  PopupPortalContainerProvider,
  UDStylesheetContainerProvider,
  usePopupPortalContainer,
} from '@universe-design/react-provider';
import ReactDOM from 'react-dom';
import { SquircleAvatar } from '../../avatar/squircle';
export default {
  title: 'Components/ConfigProvider/临时示例',
  component: ConfigProvider,
} satisfies Meta;

export const DesktopConfig = () => (
  <ConfigProvider {...DESKTOP_DEFAULT_CONFIG}>
    <Button onClick={() => toast.success(`hello random: ${Math.random()}`)}>Show Toast</Button>
  </ConfigProvider>
);

const Child = () => {
  useStyleSheetOnce('.child {background: red;}');
  return <div className="child">Child</div>;
};

export const StylesheetContainer = () => {
  const pRef = React.useRef<HTMLParagraphElement>(null);

  React.useEffect(() => {
    const rootElement = document.createElement('div');
    const container = document.createElement('div');
    const reactRoot = document.createElement('div');
    reactRoot.id = 'react-dom-root';
    container.appendChild(reactRoot);
    const shadowRoot = rootElement.attachShadow({ mode: 'open' });
    container.id = 'reactRoot';
    shadowRoot.appendChild(container);
    pRef.current?.parentElement?.appendChild(rootElement);

    ReactDOM.render(
      <React.StrictMode>
        <UDStylesheetContainerProvider container={container}>
          <div style={{ border: '6px solid #000', padding: 8 }}>
            <h1>Shadow Root!</h1>
            <Child />
            <Avatar.Group spacing={8} size={4} maxCount={2}>
              {[1, 2, 3, 4, 5].map(o => (
                <SquircleAvatar size={32} bordered key={o}>
                  {o}
                </SquircleAvatar>
              ))}
            </Avatar.Group>{' '}
          </div>
        </UDStylesheetContainerProvider>
      </React.StrictMode>,
      reactRoot,
    );
  }, []);

  return <p ref={pRef} />;
};

export const SwitchStylesheetContainer = () => {
  const [v, setV] = React.useState(true);
  return (
    <UDStylesheetContainerProvider container={v ? document.head : document.body}>
      <button
        onClick={() => {
          setV(!v);
        }}
      >
        attach to head: {v ? 'true' : 'false'}
      </button>
      <Child />
    </UDStylesheetContainerProvider>
  );
};

export const GetPopupContainer = () => {
  const getPopupContainer01 = () => document.querySelector('.container-01') as HTMLElement;
  const getPopupContainer02 = () => document.querySelector('.container-02') as HTMLElement;

  function PrintContainer() {
    const currentContainer = usePopupPortalContainer();
    const [tag, setTag] = React.useState('');
    React.useEffect(() => {
      setTag(currentContainer().tagName + '/' + currentContainer().className);
    }, [currentContainer]);
    return <p style={{ color: 'red' }}>{tag}</p>;
  }
  return (
    <div>
      <div className="main">
        <div className="container-01" />
        <div className="container-02" />
      </div>
      <ConfigProvider>
        <p>默认的的 container </p>
        <PrintContainer />
        <ConfigProvider getPopupContainer={getPopupContainer01}>
          <p>可以通过 configProvider 重置 container </p>
          <PrintContainer />
          <ConfigProvider>
            <p>继承最近的 container </p>
            <PrintContainer />
            <PopupPortalContainerProvider value={getPopupContainer02}>
              <p>可以通过 PopupPortalContainerProvider 重置 container </p>
              <PrintContainer />
            </PopupPortalContainerProvider>
          </ConfigProvider>
        </ConfigProvider>
      </ConfigProvider>
      <PopupPortalContainerProvider value={getPopupContainer02}>
        <ConfigProvider>
          <p>可以通过 PopupPortalContainerProvider 重置 container </p>
          <PrintContainer />
        </ConfigProvider>
      </PopupPortalContainerProvider>
    </div>
  );
};
