// @universe-design/react/src/config-provider/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Basic from '../demo/0.basic';
import C1TextOverflowPopperSidecar from '../demo/1.textOverflowPopperSidecar';
import C2PopperSidecar from '../demo/2.popper-sidecar';
import C3StyleSheetContainer from '../demo/3.styleSheetContainer';
if (!(C0Basic as Story).storyName) {
  (C0Basic as Story).storyName = '0. 基础用法';
  (C0Basic as Story).parameters = {
    docs: {
      description: {
        story: '以分页为例。\n\n',
      },
    },
  };
}
if (!(C1TextOverflowPopperSidecar as Story).storyName) {
  (C1TextOverflowPopperSidecar as Story).storyName = '1. 自定义文字溢出弹出组件';
  (C1TextOverflowPopperSidecar as Story).parameters = {
    docs: {
      description: {
        story:
          '可以设置 `textOverflowPopperSidecar` 来自定义文字溢出弹出组件。该属性的默认是 `Tooltip`，但是在 EE 目前的规范中，会使用 `Popover` 作为文字溢出的弹出组件。\n\n',
      },
    },
  };
}
if (!(C2PopperSidecar as Story).storyName) {
  (C2PopperSidecar as Story).storyName = '2. 使用功能更强大的 Popper 组件';
  (C2PopperSidecar as Story).parameters = {
    docs: {
      description: {
        story:
          '_2.19.0_ 版本之前我们使用 **[rc-trigger](https://github.com/react-component/trigger)** 提供组件库的弹出能力，由于其底层定位库 **[dom-align](https://github.com/yiminghe/dom-align)** 部分逻辑不符合 UD 设计规范，并且无法定制，经过调研我们决定将弹出能力切换到 **[floating-ui](https://github.com/floating-ui/floating-ui)**，感谢 rc-trigger、dom-align 和 floating-ui，更多细节可以[查阅](https://bytedance.feishu.cn/wiki/wikcnG3DTElZpfwQqB88ArZQiKe)。\n\n如果需要使用新版本 `Popper`，可以通过 `<ConfigProvider PopperSidecar={FloatingUIPopperSidecar} />` 设置。有如下几点需要注意，\n\n1. _2.19.0_ 版本之后如果使用旧版 `RcTriggerPopperSidecar` 会收到废弃警告 ⚠️。\n2. 弹层类型组件一些新增属性只有 `FloatingUIPopperSidecar` 支持，属性文档中会有标记。\n3. 目前涉及到的组件有：`Dropdown`、`Tooltip`、`Popover`、`Popconfirm`、`Select`、`TreeSelect`、`Cascader`。\n\n',
      },
    },
  };
}
if (!(C3StyleSheetContainer as Story).storyName) {
  (C3StyleSheetContainer as Story).storyName = '2. StyleSheetContainer';
  (C3StyleSheetContainer as Story).parameters = {
    docs: {
      description: {
        story:
          'UD 样式以静态 less/css 文件引入为主, 少部分样式在运行时插入。\n\n当用户在 shadowDOM 里使用这些含样式注入的组件时，由于样式注入到了外部，会使得 shadowDOM 里面的组件缺少样式。\n\n你可以使用 `UDStylesheetContainerProvider` 来指定运行时样式插入位置，来解决这个问题：\n\n',
      },
    },
  };
}

export default {
  title: 'Components/ConfigProvider/官网示例',
  tags: ['autodocs'],
};

export { C0Basic, C1TextOverflowPopperSidecar, C2PopperSidecar, C3StyleSheetContainer };
