// @universe-design/react/lib/progress/Circle.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _SucceedFilled = _interopRequireDefault(require("@universe-design/icons-react/SucceedFilled"));
var _ErrorFilled = _interopRequireDefault(require("@universe-design/icons-react/ErrorFilled"));
var _hooks = require("../config-provider/hooks");
var _clsx = _interopRequireDefault(require("clsx"));
var _shared = require("./shared");
var _util = require("../util");
function getPathStyles(percent, width, strokeWidth) {
  const radius = (width - strokeWidth) / 2;
  const len = Math.PI * 2 * radius;
  const gapDegree = 0;
  return {
    pathString: `
  M ${width / 2} ${width / 2}
  m 0 ${-radius}

  A ${radius} ${radius} 0 0 1 ${width / 2} ${width - strokeWidth / 2}
  A ${radius} ${radius} 0 0 1 ${width / 2} ${strokeWidth / 2}
  `,
    pathStyle: {
      strokeDasharray: `${percent / 100 * (len - gapDegree)}px ${len}px`,
      strokeDashoffset: `-${gapDegree / 2}px`
    }
  };
}
const Circle = props => {
  const {
    percent,
    showInfo = true,
    infoPosition = 'left',
    strokeWidth = 2,
    width = 16,
    status,
    strokeColor,
    trailColor,
    format
  } = props;
  const cssScope = (0, _hooks.useCssScope)('progress');
  const [success, exception, normal] = _shared.ProgressStatusEnum;
  const circleStatus = status || normal;
  const {
    pathString,
    pathStyle
  } = getPathStyles(percent, width, strokeWidth);
  const renderPercent = () => {
    if ((0, _util.isFunction)(format)) {
      return format(percent);
    }
    return `${percent}%`;
  };
  const renderSteps = () => {
    const cls = (0, _clsx.default)([cssScope('--circle--steps'), {
      [cssScope('--circle--steps--left')]: infoPosition === 'left',
      [cssScope('--circle--steps--right')]: infoPosition === 'right'
    }]);
    return /*#__PURE__*/_react.default.createElement("span", {
      className: cls
    }, renderPercent());
  };
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, showInfo && infoPosition === 'left' && renderSteps(), (() => /*#__PURE__*/_react.default.createElement("span", {
    style: {
      position: 'relative',
      display: 'flex'
    }
  }, /*#__PURE__*/_react.default.createElement("svg", {
    width: width,
    height: width,
    fill: "none",
    color: strokeColor,
    viewBox: `0 0 ${width} ${width}`
  }, /*#__PURE__*/_react.default.createElement("path", {
    d: pathString,
    fill: "none",
    stroke: trailColor,
    strokeWidth: strokeWidth,
    strokeLinecap: "round",
    className: (0, _clsx.default)([cssScope('--circle--path--bg'), {
      [`${cssScope('--circle--path--success')}`]: circleStatus === success,
      [`${cssScope('--circle--path--exception')}`]: circleStatus === exception
    }])
  }), /*#__PURE__*/_react.default.createElement("path", {
    d: pathString,
    fill: "none",
    stroke: strokeColor,
    strokeLinecap: percent === 0 ? 'butt' : 'round',
    strokeWidth: strokeWidth,
    style: pathStyle,
    className: (0, _clsx.default)([cssScope('--circle--path'), {
      [`${cssScope('--circle--path--success')}`]: circleStatus === success,
      [`${cssScope('--circle--path--exception')}`]: circleStatus === exception
    }])
  })), /*#__PURE__*/_react.default.createElement("span", {
    className: (0, _clsx.default)([cssScope('--circle--icon'), cssScope('--circle--icon--success')]),
    style: {
      opacity: circleStatus === success ? 1 : 0,
      fontSize: `${width}px`
    }
  }, /*#__PURE__*/_react.default.createElement(_SucceedFilled.default, {
    style: {
      verticalAlign: '10px'
    }
  })), /*#__PURE__*/_react.default.createElement("span", {
    className: (0, _clsx.default)([cssScope('--circle--icon'), cssScope('--circle--icon--exception')]),
    style: {
      opacity: circleStatus === exception ? 1 : 0,
      fontSize: `${width}px`
    }
  }, /*#__PURE__*/_react.default.createElement(_ErrorFilled.default, {
    style: {
      verticalAlign: 0
    }
  }))))(), showInfo && infoPosition === 'right' && renderSteps());
};
var _default = exports.default = Circle;
