// @universe-design/react/lib/progress/Line.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _hooks = require("../config-provider/hooks");
var _clsx = _interopRequireDefault(require("clsx"));
var _Steps = _interopRequireDefault(require("./Steps"));
const Line = props => {
  const {
    percent,
    strokeColor = '',
    strokeWidth = 4,
    status,
    trailColor,
    showInfo = true,
    infoPosition = 'right',
    format
  } = props;
  let lineStatus = status || 'normal';
  if (lineStatus === 'normal' && percent >= 100) {
    lineStatus = 'success';
  }
  const cssScope = (0, _hooks.useCssScope)('progress');
  const renderSteps = () => /*#__PURE__*/_react.default.createElement("span", {
    className: cssScope('--steps')
  }, /*#__PURE__*/_react.default.createElement(_Steps.default, {
    percent: percent,
    status: lineStatus,
    format: format
  }));
  const clsBg = (0, _clsx.default)([cssScope(), cssScope('--line-bg'), cssScope(`--line-bg-${lineStatus}`)]);
  const clsInnerLine = (0, _clsx.default)([cssScope(), cssScope('--line-inner'), {
    [cssScope(`--line-inner-with-${infoPosition}-info`)]: showInfo
  }]);
  const innerStyle = {
    height: strokeWidth
  };
  if (trailColor) {
    innerStyle.backgroundColor = trailColor;
  }
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, showInfo && infoPosition === 'left' && renderSteps(), /*#__PURE__*/_react.default.createElement("div", {
    className: clsInnerLine,
    style: innerStyle
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: clsBg,
    style: {
      width: `${percent}%`,
      [strokeColor && 'backgroundColor']: strokeColor
    }
  })), showInfo && infoPosition === 'right' && renderSteps());
};
var _default = exports.default = Line;
