// @universe-design/styles/ts/im/index.light.atom.ts
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
export const im_btn_selected_bg_gray_hover = 'rgba(31, 35, 41, 0.1)';
export const im_btn_selected_bg_gray_normal = 'rgba(31, 35, 41, 0.06)';
export const im_btn_selected_bg_gray_press = 'rgba(31, 35, 41, 0.14)';
export const im_btn_selected_bg_green_hover = '#d0f5ce';
export const im_btn_selected_bg_green_normal = '#e4fae1';
export const im_btn_selected_bg_green_press = '#95e599';
export const im_btn_selected_bg_red_hover = '#fee3e2';
export const im_btn_selected_bg_red_normal = '#fef0f0';
export const im_btn_selected_bg_red_press = '#fdc6c4';
export const im_btn_selected_line_gray = '#646a73';
export const im_btn_selected_text_green = '#1a7526';
export const im_btn_selected_text_red = '#e22e28';
export const imtoken_AI_Profile_Avatar_bg = 'linear-gradient(to top, #ffffff 32%, #e1d6f9 66%, #b780e0 100%)';
export const imtoken_AI_message_bg = '#f4f5f8';
export const imtoken_AI_onboarding_Avatar_bg = 'rgba(0, 0, 0, 0)';
export const imtoken_add_menu_icon_bg = '#ffffff';
export const imtoken_bg_Aurora_Horizon = '#ecedee';
export const imtoken_dialog_bg_top = 'linear-gradient(180deg, #e4f2ff 0%, #ffffff 100%)';
export const imtoken_dialog_bg_top_cricle =
  'linear-gradient(180deg, rgba(211, 229, 255, 0.64) 0%, rgba(255, 255, 255, 0.8) 100%)';
export const imtoken_feed_bg_body = '#ffffff';
export const imtoken_feed_bg_complete = '#10a893';
export const imtoken_feed_bg_delete = '#e22e28';
export const imtoken_feed_bg_handle_later = '#fad355';
export const imtoken_feed_bg_reaction = 'rgba(31, 35, 41, 0.1)';
export const imtoken_feed_bg_top = '#646a73';
export const imtoken_feed_fill_active = 'rgba(51, 133, 255, 0.12)';
export const imtoken_feed_icon_pri_selected = '#1456f0';
export const imtoken_feed_item_active_btn_default = 'rgba(31, 35, 41, 0.08)';
export const imtoken_feed_item_active_btn_hover = 'rgba(31, 35, 41, 0.13)';
export const imtoken_feed_item_active_btn_press = 'rgba(31, 35, 41, 0.18)';
export const imtoken_feed_item_bg_press = 'rgba(31, 35, 41, 0.12)';
export const imtoken_feed_item_default_btn_default = '#eff0f1';
export const imtoken_feed_item_hover_btn_default = '#d8d9db';
export const imtoken_feed_item_hover_btn_hover = '#d0d1d3';
export const imtoken_feed_item_hover_btn_press = '#cacbcd';
export const imtoken_feed_text_pri_selected = '#1456f0';
export const imtoken_feed_topitem_active_btn_default = '#ffffff';
export const imtoken_feed_topitem_active_btn_hover = 'rgba(255, 255, 255, 0.65)';
export const imtoken_feed_topitem_active_btn_press = 'rgba(31, 35, 41, 0.05)';
export const imtoken_feed_topitem_bg_default = '#f2f3f5';
export const imtoken_feed_topitem_bg_hover = '#e8e9eb';
export const imtoken_feed_topitem_bg_press = '#e1e2e4';
export const imtoken_feed_topitem_default_btn_default = '#ffffff';
export const imtoken_feed_topitem_hover_btn_default = '#ffffff';
export const imtoken_feed_topitem_hover_btn_hover = 'rgba(255, 255, 255, 0.6)';
export const imtoken_feed_topitem_hover_btn_press = 'rgba(31, 35, 41, 0.05)';
export const imtoken_general_icon_encrypted_chat_normal = '#373c43';
export const imtoken_message_bg_audio_blue = 'rgba(20, 86, 240, 0.15)';
export const imtoken_message_bg_audio_grey = 'rgba(31, 35, 41, 0.05)';
export const imtoken_message_bg_bubbles_blue = '#d1e3ff';
export const imtoken_message_bg_bubbles_grey = '#ebeced';
export const imtoken_message_bg_chat = '#fcfcfc';
export const imtoken_message_bg_chat_simple = '#fcfcfc';
export const imtoken_message_bg_location = '#faeec4';
export const imtoken_message_bg_pin = '#fcf5df';
export const imtoken_message_bg_reaction_blue = 'rgba(20, 86, 240, 0.15)';
export const imtoken_message_bg_reaction_grey = 'rgba(31, 35, 41, 0.05)';
export const imtoken_message_bg_reaction_grey_body = 'rgba(31, 35, 41, 0.05)';
export const imtoken_message_bg_reaction_grey_simple = 'rgba(31, 35, 41, 0.05)';
export const imtoken_message_bg_selected = 'rgba(0, 0, 0, 0.02)';
export const imtoken_message_card_bg_body = '#ffffff';
export const imtoken_message_card_bg_body_embed = '#ffffff';
export const imtoken_message_card_border = '#dee0e3';
export const imtoken_message_icon_lan = '#ff811a';
export const imtoken_message_icon_pin = '#10a893';
export const imtoken_message_icon_pin_list = '#10a893';
export const imtoken_message_icon_read = '#10a893';
export const imtoken_message_icon_unread = '#8f959e';
export const imtoken_message_icon_voice_blue = '#000000';
export const imtoken_message_icon_voice_blue_bg = '#0442d2';
export const imtoken_message_icon_voice_grey = '#000000';
export const imtoken_message_icon_voice_grey_bg = '#373c43';
export const imtoken_message_pin_list_popover_caret = '#fff8f0';
export const imtoken_message_pin_popover_bg = '#f0f4ff';
export const imtoken_message_pin_popover_line = '#ffffff';
export const imtoken_message_pin_text_bg_focus = 'rgba(15, 133, 117, 0.1)';
export const imtoken_message_pin_text_bg_hover = 'rgba(15, 133, 117, 0.1)';
export const imtoken_message_pin_text_bg_pressed = 'rgba(15, 133, 117, 0.2)';
export const imtoken_message_sidebar_group_bg = '#f5f6f7';
export const imtoken_message_sidebar_group_bg_simple = '#f5f6f7';
export const imtoken_message_text_bubbles_blue = '#1456f0';
export const imtoken_message_text_bubbles_grey = '#51565d';
export const imtoken_message_text_pin = '#10a893';
export const imtoken_message_voice_line_schedule_blue = '#0442d2';
export const imtoken_message_voice_line_schedule_grey = '#373c43';
export const imtoken_message_voice_text_time_blue = '#0442d2';
export const imtoken_message_voice_text_time_gray = '#1f2329';
export const imtoken_profile_bg_hover = '#eff0f1';
export const imtoken_profile_bg_normal = '#f5f6f7';
export const imtoken_profile_bg_press = '#eff0f1';
export const imtoken_profile_icon_normal = '#373c43';
export const imtoken_profile_text_normal = '#2b2f36';
export const imtoken_red_envelope_button_text_yellow = '#faedc2';
export const imtoken_red_envelope_card_text_placeholder_yellow = 'rgba(252, 223, 126, 0.8)';
export const imtoken_red_envelope_card_text_title_yellow = '#fcdf7e';
export const imtoken_secret_chat_bg_topbar = '#383b42';
export const imtoken_secure_badge_bg_normal = '#373c43';
export const imtoken_set_colour_Aurora_Air = 'linear-gradient(45deg, #e3ecf7 0%, #d5dff6 100%)';
export const imtoken_set_colour_Aurora_Horizon = '#ecedee';
export const imtoken_set_colour_Aurora_Lake = 'linear-gradient(45deg, #476dae 0%, #4977ab 100%)';
export const imtoken_set_colour_Aurora_Obsidian = 'linear-gradient(45deg, #3f4b64 0%, #697c96 100%)';
export const imtoken_set_colour_Aurora_Sky = '#e6f1f4';
export const imtoken_side_unread_bg_active = '#dee8f8';
export const imtoken_side_unread_bg_default = '#f5f6f7';
export const imtoken_side_unread_bg_hover = '#e4e5e7';
export const imtoken_t_entrance_default = 'rgba(255, 255, 255, 0.7)';
export const imtoken_t_entrance_hover = 'rgba(255, 255, 255, 0.9)';
export const imtoken_text_tips_link_hover = '#336df4';
export const imtoken_text_tips_link_normal = '#5083fb';
export const imtoken_text_tips_link_pressed = '#94b4ff';
export const imtoken_topic_group_comment_file = '#ffffff';
