// @universe-design/react/src/hooks/useAsyncFn.ts
import * as React from 'react';
import type { FunctionReturningPromise, PromiseType } from './types';

export type AsyncState<T> =
  | {
      loading: true;
      error?: Error | undefined;
      value?: T;
    }
  | {
      loading: false;
      error: Error;
      value?: undefined;
    }
  | {
      loading: false;
      error?: undefined;
      value: T;
    };

type StateFromFunctionReturningPromise<T extends FunctionReturningPromise> = AsyncState<PromiseType<ReturnType<T>>>;

export type AsyncFnReturn<T extends FunctionReturningPromise = FunctionReturningPromise> = [
  StateFromFunctionReturningPromise<T>,
  T,
];

const INITIAL_STATE = { loading: false, error: new Error() };

export default function useAsyncFn<T extends FunctionReturningPromise>(
  fn: T,
  {
    initialState = INITIAL_STATE,
  }: {
    initialState?: StateFromFunctionReturningPromise<T>;
  } = {},
): AsyncFnReturn<T> {
  const lastCallId = React.useRef(0);
  const [state, setState] = React.useState<StateFromFunctionReturningPromise<T>>(initialState);

  const callback = React.useCallback(
    (...args: Parameters<T>): ReturnType<T> => {
      const callId = ++lastCallId.current;

      setState(prevState => ({ ...prevState, loading: true }));

      return fn(...args).then(
        value => {
          callId === lastCallId.current &&
            setState({
              value: value as PromiseType<ReturnType<T>>,
              loading: false,
            });
          return value;
        },
        error => {
          callId === lastCallId.current && setState({ error, loading: false });
          return error;
        },
      ) as ReturnType<T>;
    },
    [fn],
  );

  return [state, callback as unknown as T];
}
