// @universe-design/react/src/hooks/useControlled.ts
import type React from 'react';
import { useState, useCallback } from 'react';

// 内部 state 改变，要触发 props.onChange(state)
// 外部 value 改变，要设置内部 state

export interface Option<T> {
  defaultValue?: T;
  defaultValuePropName?: string;
  valuePropName?: string;
  onChangePropName?: string;
}

export interface Props {
  [key: string]: any;
}

export default <T>(props: Props = {}, options: Option<T> = {}): [T, React.Dispatch<React.SetStateAction<T>>] => {
  const {
    defaultValue: _defaultValue,
    defaultValuePropName = 'defaultValue',
    valuePropName = 'value',
    onChangePropName = 'onChange',
  } = options;
  const propValue = props[valuePropName];
  const propOnChange = props[onChangePropName];
  const defaultValue = defaultValuePropName in props ? props[defaultValuePropName] : _defaultValue;
  const isControlled = valuePropName in props;
  const [innerState, setInnerState] = useState<T>(() => (isControlled ? propValue : defaultValue));
  const state = (isControlled ? propValue : innerState) as T;

  const setState = useCallback(
    _state => {
      setInnerState(_state);
      if (propOnChange) {
        propOnChange(_state);
      }
    },
    [propOnChange],
  );
  return [state, setState];
};
