// @universe-design/react/src/hooks/useComplementEvent.ts
import type { VirtualElement } from '@floating-ui/core';
import * as React from 'react';
import { isElement } from '@universe-design/react-utilities';
import type { HookSharedOptions } from '../util';
import { isNonNullable } from '../util';
import { contains } from '../util/dom';
import type { ThunkElement } from '../util/getThunkElement';
import { getThunkElement } from '../util/getThunkElement';

export const getElementFromRefOrVirtualElement = <T extends Node>(
  elementOrRef: ThunkElement<T | VirtualElement>,
): T | null => {
  const element = getThunkElement(elementOrRef);
  if (!element) {
    return element;
  }
  if ('contextElement' in element) {
    return element.contextElement as T;
  }
  if (isElement(element)) {
    return element as unknown as T;
  }
  return null;
};

// TODO 这里可能会是一个性能瓶颈点，因为每个 popper 元素都会做一次 document 的事件绑定
export function useComplementEvent<K extends keyof DocumentEventMap>(
  type: K,
  {
    listener,
    elements,
    options,
    skipped = false,
  }: {
    listener: (ev: DocumentEventMap[K]) => any;
    elements: ThunkElement<Node | VirtualElement>[];
    options?: boolean | AddEventListenerOptions;
  } & HookSharedOptions,
) {
  React.useEffect(() => {
    if (skipped) {
      return;
    }
    function handleDocumentEvent(ev: DocumentEventMap[K]): any {
      const { target } = ev;
      // console.trace('useComplementEvent', {
      //   elements: elements.map(getElementFromRefOrVirtualElement),
      //   target,
      // });
      if (
        (target instanceof HTMLElement || target instanceof SVGElement) &&
        // 点击的元素如果不是我们的目标 elements，或者不在我们的目标 elements 内部
        elements
          .map(getElementFromRefOrVirtualElement)
          .filter(isNonNullable)
          .every(element => {
            if (ev.composedPath) {
              // 处理 shadowRoot 情况
              return !(ev.composedPath().indexOf(element) !== -1);
            }
            return !contains(element, target);
          })
      ) {
        // console.trace('useComplementEvent call listener', {
        //   elements: elements.map(getElementFromRefOrVirtualElement),
        //   target,
        // });
        listener(ev);
      }
    }

    document.addEventListener(type, handleDocumentEvent, options);
    return () => {
      document.removeEventListener(type, handleDocumentEvent, options);
    };
  }, [elements, listener, options, skipped, type]);
}
