// @universe-design/react/src/hooks/useObservable.ts
/* eslint-disable no-redeclare */
import isEqual from 'lodash/isEqual';
import { useCallback, useMemo, useRef, useState } from 'react';
import type { Subscription } from 'rxjs';
import { Subject, timer } from 'rxjs';
import { debounce, map } from 'rxjs/operators';
import { useConstant } from '@universe-design/react-utilities';
import { useEffectOnce } from './useEffectOnce';
import { useIsomorphicLayoutEffect as useLayoutEffect } from './useIsomorphicLayoutEffect';

export interface Observable<T> {
  subscribe: (listener: (value: T) => void) => {
    unsubscribe: () => void;
  };
}

export function useObservable<T>(observable$: Observable<T>): T | undefined;
export function useObservable<T>(observable$: Observable<T>, initialValue: T): T;
export function useObservable<T>(observable$: Observable<T>, initialValue?: T): T | undefined {
  const [value, update] = useState<T | undefined>(initialValue);

  useLayoutEffect(() => {
    const s = observable$.subscribe(update);
    return () => s.unsubscribe();
  }, [observable$]);

  return value;
}

export function useDebouncedCallback<T extends (...args: never[]) => unknown>(fn: T, ms = 0) {
  const event$ = useConstant(() => new Subject<Parameters<T>>());
  useEffectOnce(() => () => {
    event$.complete();
  });
  const value$ = useMemo(
    () =>
      event$.pipe(
        debounce(() => timer(ms)),
        map(args => fn(...args) as ReturnType<T>),
      ),
    [event$, fn, ms],
  );
  const s = useRef<Subscription>();
  useLayoutEffect(() => {
    s.current = value$.subscribe();
    return () => s.current?.unsubscribe();
  }, [value$]);
  return useMemo(() => {
    const func: {
      (...args: Parameters<T>): void;
      cancel: () => void;
    } = (...args: Parameters<T>) => {
      event$.next(args);
    };
    func.cancel = () => {
      s.current?.unsubscribe();
    };
    return func;
  }, [event$]);
}

export function useDebounce<T>(
  value: T,
  delay: number,
  options?: {
    equalityFn?: (left: T, right: T) => boolean;
  },
) {
  const eq = options?.equalityFn || isEqual;

  const [state, setState] = useState(value);
  const [preState, setPreState] = useState<T>();
  const debounced = useDebouncedCallback(
    useCallback((_value: T) => setState(_value), []),
    delay,
  );

  if (preState === undefined || !eq(preState, value)) {
    debounced(value);
    setPreState(value);
  }

  return [
    state,
    {
      cancel: debounced.cancel,
    },
  ] as const;
}
