// @universe-design/react/src/hooks/useScrollStatus.tsx
import { useCallback } from 'react';
import { useImmer } from 'use-immer';
import type { ThunkElement } from '../util/getThunkElement';
import { getThunkElement } from '../util/getThunkElement';

/**
 * 检测滚动容器的滚动状态
 */
export function useScrollStatus<T extends HTMLElement = HTMLDivElement>(
  params: {
    /** 禁用水平滚动监听 */
    disableHorizontal?: boolean;
    /** 禁用竖直滚动监听 */
    disableVertical?: boolean;
  } = {},
) {
  const { disableHorizontal = false, disableVertical = false } = params;

  const [overflow, updateOverflow] = useImmer({
    t: false,
    b: false,
    l: false,
    r: false,
  });

  const checkOverflow = useCallback(
    (target: ThunkElement<T>) => {
      const targetEl = getThunkElement(target);
      if (!targetEl) {
        return;
      }

      const { scrollTop, scrollHeight, clientHeight, scrollLeft, scrollWidth, clientWidth } = targetEl;

      const targetInset = {
        top: 0,
        bottom: clientHeight,
        left: 0,
        right: clientWidth,
      };

      const childInset = {
        top: -scrollTop,
        bottom: -scrollTop + scrollHeight,
        left: -scrollLeft,
        right: -scrollLeft + scrollWidth,
      };

      updateOverflow(prevOverflow => ({
        t: disableVertical ? prevOverflow.t : targetInset.top > childInset.top,
        b: disableVertical ? prevOverflow.b : targetInset.bottom < childInset.bottom,
        l: disableHorizontal ? prevOverflow.l : targetInset.left > childInset.left,
        r: disableHorizontal ? prevOverflow.r : targetInset.right < childInset.right,
      }));
    },
    [disableHorizontal, disableVertical, updateOverflow],
  );

  return {
    checkOverflow,
    overflow,
  };
}
