// @universe-design/illustration-react/dist/cjs/extraComponents/loading/EmptyPositiveLoadingBase.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
require('./style/index.css');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n["default"] = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespace(React);

function LoadingLight(props) {
    const { size = 'normal', svgSource } = props;
    let loadingSize;
    if (size === 'normal') {
        loadingSize = 120;
    }
    else if (size === 'small') {
        loadingSize = 100;
    }
    else {
        loadingSize = size;
    }
    loadingSize = typeof loadingSize === 'number' ? `${loadingSize}px` : loadingSize;
    const style = {
        ...props.style,
        width: loadingSize,
        height: loadingSize,
        overflow: "hidden",
        backgroundSize: `auto ${loadingSize}`,
        backgroundImage: `url(${svgSource})`
    };
    const classNames = `ud__illustration__loading-container ` + (props.classNames || '');
    return (React__namespace.createElement("div", { className: classNames, style: style }));
}

exports["default"] = LoadingLight;
