// @universe-design/styles/ts/preset-light.atom.ts
// autogen by ../scripts/gen-ts-from-json.ts
export const AI_cursor = 'rgb(71, 82, 230)';
export const AI_dynamic_line =
  'conic-gradient(from 45deg, rgb(207, 94, 207) 0%, rgb(71, 82, 230) 33.33333333333333%, rgb(71, 82, 230) 66.66666666666666%, rgb(207, 94, 207) 100%)';
export const AI_icon_gradient = 'linear-gradient(to right, rgb(71, 82, 230) 0%, rgb(159, 111, 241) 100%)';
export const AI_icon_in_button = 'rgb(71, 82, 230)';
export const AI_loading = 'linear-gradient(to right, rgb(71, 82, 230) 0%, rgb(222, 129, 222) 100%)';
export const AI_primary_content_default = 'linear-gradient(to right, rgb(71, 82, 230) 0%, rgb(207, 94, 207) 100%)';
export const AI_primary_content_hover = 'linear-gradient(to right, rgb(91, 101, 245) 0%, rgb(222, 129, 222) 100%)';
export const AI_primary_content_loading =
  'linear-gradient(to right, rgba(71, 82, 230, 0.6) 0%, rgba(207, 94, 207, 0.6) 100%)';
export const AI_primary_content_pressed = 'linear-gradient(to right, rgb(71, 82, 230) 0%, rgb(191, 61, 191) 100%)';
export const AI_primary_fill_default = 'linear-gradient(to right, rgb(91, 101, 245) 0%, rgb(222, 129, 222) 100%)';
export const AI_primary_fill_hover = 'linear-gradient(to right, rgb(117, 125, 240) 0%, rgb(229, 156, 229) 100%)';
export const AI_primary_fill_loading =
  'linear-gradient(to right, rgba(91, 101, 245, 0.6) 0%, rgba(222, 129, 222, 0.6) 100%)';
export const AI_primary_fill_pressed = 'linear-gradient(to right, rgb(71, 82, 230) 0%, rgb(207, 94, 207) 100%)';
export const AI_primary_fill_solid_01 = 'linear-gradient(to right, rgb(242, 243, 253) 0%, rgb(252, 238, 252) 100%)';
export const AI_primary_fill_solid_02 = 'linear-gradient(to right, rgb(233, 234, 251) 0%, rgb(249, 226, 249) 100%)';
export const AI_primary_fill_solid_03 = 'linear-gradient(to right, rgb(204, 207, 249) 0%, rgb(243, 196, 243) 100%)';
export const AI_primary_fill_transparent_01 =
  'linear-gradient(to right, rgba(91, 101, 245, 0.08) 0%, rgba(222, 129, 222, 0.08) 100%)';
export const AI_primary_fill_transparent_02 =
  'linear-gradient(to right, rgba(91, 101, 245, 0.2) 0%, rgba(222, 129, 222, 0.2) 100%)';
export const AI_primary_fill_transparent_03 =
  'linear-gradient(to right, rgba(91, 101, 245, 0.3) 0%, rgba(222, 129, 222, 0.3) 100%)';
export const AI_sendicon = 'linear-gradient(to right, rgb(71, 82, 230) 0%, rgb(159, 111, 241) 100%)';
export const Aurora_bg_main = 'rgba(20, 86, 240, 0.2)';
export const Aurora_bg_reflection = 'rgba(45, 190, 171, 0.1)';
export const Aurora_bg_subordinate = 'rgba(51, 109, 244, 0.2)';
export const B100 = '#e0e9ff';
export const B200 = '#c2d4ff';
export const B300 = '#94b4ff';
export const B350 = '#7aa2ff';
export const B400 = '#5083fb';
export const B50 = '#f0f4ff';
export const B500 = '#336df4';
export const B600 = '#1456f0';
export const B700 = '#0442d2';
export const B800 = '#002f9e';
export const B900 = '#002270';
export const C100 = '#fee2f2';
export const C200 = '#f8c4e1';
export const C300 = '#f598cc';
export const C350 = '#eb78b8';
export const C400 = '#df58a5';
export const C50 = '#fef0f8';
export const C500 = '#cc398c';
export const C600 = '#b82879';
export const C700 = '#9d1562';
export const C800 = '#730744';
export const C900 = '#550c35';
export const G100 = '#d0f5ce';
export const G200 = '#95e599';
export const G300 = '#5cd168';
export const G350 = '#35bd4b';
export const G400 = '#32a645';
export const G50 = '#e4fae1';
export const G500 = '#258832';
export const G600 = '#1a7526';
export const G700 = '#0b6017';
export const G800 = '#04430c';
export const G900 = '#022c07';
export const I100 = '#e9eafb';
export const I200 = '#cccff9';
export const I300 = '#abb0f2';
export const I350 = '#959bee';
export const I400 = '#757df0';
export const I50 = '#f2f3fd';
export const I500 = '#5b65f5';
export const I600 = '#4752e6';
export const I700 = '#333dcc';
export const I800 = '#1e27a4';
export const I900 = '#151b70';
export const L100 = '#e3f0a3';
export const L200 = '#c8dd5f';
export const L300 = '#a2c10b';
export const L350 = '#91ad00';
export const L400 = '#7b9207';
export const L50 = '#f2f8d3';
export const L500 = '#6b7f06';
export const L600 = '#5c6d08';
export const L700 = '#4a5804';
export const L800 = '#333d00';
export const L900 = '#262e00';
export const N00 = '#ffffff';
export const N00_10 = 'rgba(255, 255, 255, 0.1)';
export const N00_15 = 'rgba(255, 255, 255, 0.15)';
export const N00_20 = 'rgba(255, 255, 255, 0.2)';
export const N00_30 = 'rgba(255, 255, 255, 0.3)';
export const N00_40 = 'rgba(255, 255, 255, 0.4)';
export const N00_5 = 'rgba(255, 255, 255, 0.05)';
export const N00_50 = 'rgba(255, 255, 255, 0.5)';
export const N00_60 = 'rgba(255, 255, 255, 0.6)';
export const N00_70 = 'rgba(255, 255, 255, 0.7)';
export const N00_80 = 'rgba(255, 255, 255, 0.8)';
export const N00_90 = 'rgba(255, 255, 255, 0.9)';
export const N100 = '#f2f3f5';
export const N1000 = '#000000';
export const N200 = '#eff0f1';
export const N300 = '#dee0e3';
export const N350 = '#d0d3d6';
export const N400 = '#bbbfc4';
export const N50 = '#f5f6f7';
export const N500 = '#8f959e';
export const N600 = '#646a73';
export const N650 = '#51565d';
export const N700 = '#373c43';
export const N800 = '#2b2f36';
export const N900 = '#1f2329';
export const N900_10 = 'rgba(31, 35, 41, 0.1)';
export const N900_15 = 'rgba(31, 35, 41, 0.15)';
export const N900_20 = 'rgba(31, 35, 41, 0.2)';
export const N900_30 = 'rgba(31, 35, 41, 0.3)';
export const N900_40 = 'rgba(31, 35, 41, 0.4)';
export const N900_5 = 'rgba(31, 35, 41, 0.05)';
export const N900_50 = 'rgba(31, 35, 41, 0.5)';
export const N900_60 = 'rgba(31, 35, 41, 0.6)';
export const N900_70 = 'rgba(31, 35, 41, 0.7)';
export const N900_80 = 'rgba(31, 35, 41, 0.8)';
export const N900_90 = 'rgba(31, 35, 41, 0.9)';
export const N950 = '#0f1114';
export const O100 = '#fee7cd';
export const O200 = '#fec48b';
export const O300 = '#ff9d4c';
export const O350 = '#ff811a';
export const O400 = '#ed6d0c';
export const O50 = '#fff3e5';
export const O500 = '#c25705';
export const O600 = '#a44904';
export const O700 = '#853a05';
export const O800 = '#642b02';
export const O900 = '#3b1a02';
export const P100 = '#efe6fe';
export const P200 = '#dcc9fd';
export const P300 = '#c8a9fc';
export const P350 = '#b791fa';
export const P400 = '#9f6ff1';
export const P50 = '#f5f0fe';
export const P500 = '#8d55ed';
export const P600 = '#7a35f0';
export const P700 = '#611fd6';
export const P800 = '#4811a6';
export const P900 = '#2f0080';
export const R100 = '#fee3e2';
export const R200 = '#fdc6c4';
export const R300 = '#f89e9b';
export const R350 = '#ff7570';
export const R400 = '#f54a45';
export const R50 = '#fef0f0';
export const R500 = '#e22e28';
export const R600 = '#c02a26';
export const R700 = '#a11c17';
export const R800 = '#741915';
export const R900 = '#590603';
export const S100 = '#fffca3';
export const S200 = '#fff67a';
export const S300 = '#fff258';
export const S350 = '#ffe928';
export const S400 = '#e5ce00';
export const S50 = '#ffffdb';
export const S500 = '#c2a800';
export const S600 = '#8f7c00';
export const S700 = '#5c4f00';
export const S800 = '#423700';
export const S900 = '#2c2502';
export const T100 = '#c4f2ec';
export const T200 = '#6fe8d8';
export const T300 = '#33d6c0';
export const T350 = '#2dbeab';
export const T400 = '#10a893';
export const T50 = '#e2f8f5';
export const T500 = '#0f8575';
export const T600 = '#067062';
export const T700 = '#045d51';
export const T800 = '#03443b';
export const T900 = '#02312a';
export const V100 = '#f9e2f9';
export const V200 = '#f3c4f3';
export const V300 = '#e59ce5';
export const V350 = '#de81de';
export const V400 = '#cf5ecf';
export const V50 = '#fceefc';
export const V500 = '#bf3dbf';
export const V600 = '#a630a6';
export const V700 = '#872787';
export const V800 = '#6a116a';
export const V900 = '#520052';
export const Vnext_bg_content_header = 'rgb(252, 252, 252)';
export const Vnext_bg_gradient =
  'linear-gradient(180deg, rgba(255, 255, 255, 0.05) 0%, rgba(100, 106, 115, 0.05) 100%)';
export const Vnext_bg_sub_navigation = 'rgba(245, 245, 245, 0.7)';
export const Vnext_bg_thirdly_navigation = 'rgba(252, 252, 252, 0.9)';
export const W100 = '#caeffc';
export const W200 = '#97dcfc';
export const W300 = '#3ec3f7';
export const W350 = '#25b0e7';
export const W400 = '#1295ca';
export const W50 = '#e7f8fe';
export const W500 = '#047fb0';
export const W600 = '#076a94';
export const W700 = '#0f587a';
export const W800 = '#06415c';
export const W900 = '#072b3d';
export const Y100 = '#faedc2';
export const Y200 = '#fcdf7e';
export const Y300 = '#fad355';
export const Y350 = '#ffc60a';
export const Y400 = '#d99904';
export const Y50 = '#fbf4df';
export const Y500 = '#ad7a03';
export const Y600 = '#865b03';
export const Y700 = '#6f4a01';
export const Y800 = '#573601';
export const Y900 = '#382201';
export const bg_base = 'rgb(242, 243, 245)';
export const bg_body = 'rgb(255, 255, 255)';
export const bg_body_overlay = 'rgb(245, 246, 247)';
export const bg_content_base = 'rgb(248, 249, 250)';
export const bg_filler = 'rgb(239, 240, 241)';
export const bg_float = 'rgb(255, 255, 255)';
export const bg_float_base = 'rgb(242, 243, 245)';
export const bg_float_overlay = 'rgb(245, 246, 247)';
export const bg_float_push = 'rgba(255, 255, 255, 0.8)';
export const bg_mask = 'rgba(0, 0, 0, 0.55)';
export const bg_pricolor = 'rgb(80, 131, 251)';
export const bg_sub_navigation = 'rgb(245, 246, 247)';
export const bg_text_selection = 'rgba(20, 86, 240, 0.3)';
export const bg_tips = 'rgb(31, 35, 41)';
export const colorful_blue = 'rgb(51, 109, 244)';
export const colorful_carmine = 'rgb(223, 88, 165)';
export const colorful_green = 'rgb(53, 189, 75)';
export const colorful_indigo = 'rgb(91, 101, 245)';
export const colorful_lime = 'rgb(145, 173, 0)';
export const colorful_neutral = 'rgb(143, 149, 158)';
export const colorful_orange = 'rgb(255, 129, 26)';
export const colorful_purple = 'rgb(141, 85, 237)';
export const colorful_red = 'rgb(245, 74, 69)';
export const colorful_sunflower = 'rgb(255, 233, 40)';
export const colorful_turquoise = 'rgb(45, 190, 171)';
export const colorful_violet = 'rgb(191, 61, 191)';
export const colorful_wathet = 'rgb(37, 176, 231)';
export const colorful_yellow = 'rgb(255, 198, 10)';
export const fill_active = 'rgba(20, 86, 240, 0.15)';
export const fill_disabled = 'rgb(187, 191, 196)';
export const fill_focus = 'rgba(31, 35, 41, 0.12)';
export const fill_hover = 'rgba(31, 35, 41, 0.08)';
export const fill_img_mask = 'rgba(255, 255, 255, 0)';
export const fill_loading_mask = 'rgba(255, 255, 255, 0.6)';
export const fill_pressed = 'rgba(31, 35, 41, 0.12)';
export const fill_selected = 'rgba(20, 86, 240, 0.1)';
export const fill_tag = 'rgba(31, 35, 41, 0.1)';
export const function_danger_100 = 'rgb(254, 227, 226)';
export const function_danger_200 = 'rgb(253, 198, 196)';
export const function_danger_300 = 'rgb(248, 158, 155)';
export const function_danger_350 = 'rgb(255, 117, 112)';
export const function_danger_400 = 'rgb(245, 74, 69)';
export const function_danger_50 = 'rgb(254, 240, 240)';
export const function_danger_500 = 'rgb(226, 46, 40)';
export const function_danger_600 = 'rgb(192, 42, 38)';
export const function_danger_700 = 'rgb(161, 28, 23)';
export const function_danger_800 = 'rgb(116, 25, 21)';
export const function_danger_900 = 'rgb(89, 6, 3)';
export const function_danger_content_default = 'rgb(245, 74, 69)';
export const function_danger_content_hover = 'rgb(255, 117, 112)';
export const function_danger_content_loading = 'rgb(248, 158, 155)';
export const function_danger_content_pressed = 'rgb(226, 46, 40)';
export const function_danger_fill_default = 'rgb(245, 74, 69)';
export const function_danger_fill_hover = 'rgb(255, 117, 112)';
export const function_danger_fill_loading = 'rgb(248, 158, 155)';
export const function_danger_fill_pressed = 'rgb(226, 46, 40)';
export const function_danger_fill_solid_01 = 'rgb(254, 240, 240)';
export const function_danger_fill_solid_02 = 'rgb(254, 227, 226)';
export const function_danger_fill_solid_03 = 'rgb(253, 198, 196)';
export const function_danger_fill_transparent_01 = 'rgba(226, 46, 40, 0.1)';
export const function_danger_fill_transparent_02 = 'rgba(226, 46, 40, 0.15)';
export const function_danger_fill_transparent_03 = 'rgba(226, 46, 40, 0.2)';
export const function_danger_on_danger_fill = 'rgb(255, 255, 255)';
export const function_info_100 = 'rgb(224, 233, 255)';
export const function_info_200 = 'rgb(194, 212, 255)';
export const function_info_300 = 'rgb(148, 180, 255)';
export const function_info_350 = 'rgb(122, 162, 255)';
export const function_info_400 = 'rgb(80, 131, 251)';
export const function_info_50 = 'rgb(240, 244, 255)';
export const function_info_500 = 'rgb(51, 109, 244)';
export const function_info_600 = 'rgb(20, 86, 240)';
export const function_info_700 = 'rgb(4, 66, 210)';
export const function_info_800 = 'rgb(0, 47, 158)';
export const function_info_900 = 'rgb(0, 34, 112)';
export const function_info_content_default = 'rgb(20, 86, 240)';
export const function_info_content_hover = 'rgb(51, 109, 244)';
export const function_info_content_loading = 'rgb(148, 180, 255)';
export const function_info_content_pressed = 'rgb(4, 66, 210)';
export const function_info_fill_default = 'rgb(20, 86, 240)';
export const function_info_fill_hover = 'rgb(51, 109, 244)';
export const function_info_fill_loading = 'rgb(148, 180, 255)';
export const function_info_fill_pressed = 'rgb(4, 66, 210)';
export const function_info_fill_solid_01 = 'rgb(240, 244, 255)';
export const function_info_fill_solid_02 = 'rgb(224, 233, 255)';
export const function_info_fill_solid_03 = 'rgb(194, 212, 255)';
export const function_info_fill_transparent_01 = 'rgba(20, 86, 240, 0.1)';
export const function_info_fill_transparent_02 = 'rgba(20, 86, 240, 0.15)';
export const function_info_fill_transparent_03 = 'rgba(20, 86, 240, 0.2)';
export const function_info_on_info_fill = 'rgb(255, 255, 255)';
export const function_success_100 = 'rgb(208, 245, 206)';
export const function_success_200 = 'rgb(149, 229, 153)';
export const function_success_300 = 'rgb(92, 209, 104)';
export const function_success_350 = 'rgb(53, 189, 75)';
export const function_success_400 = 'rgb(50, 166, 69)';
export const function_success_50 = 'rgb(228, 250, 225)';
export const function_success_500 = 'rgb(37, 136, 50)';
export const function_success_600 = 'rgb(26, 117, 38)';
export const function_success_700 = 'rgb(11, 96, 23)';
export const function_success_800 = 'rgb(4, 67, 12)';
export const function_success_900 = 'rgb(2, 44, 7)';
export const function_success_content_default = 'rgb(50, 166, 69)';
export const function_success_content_hover = 'rgb(53, 189, 75)';
export const function_success_content_loading = 'rgb(92, 209, 104)';
export const function_success_content_pressed = 'rgb(37, 136, 50)';
export const function_success_fill_default = 'rgb(53, 189, 75)';
export const function_success_fill_hover = 'rgb(92, 209, 104)';
export const function_success_fill_loading = 'rgb(92, 209, 104)';
export const function_success_fill_pressed = 'rgb(50, 166, 69)';
export const function_success_fill_solid_01 = 'rgb(228, 250, 225)';
export const function_success_fill_solid_02 = 'rgb(208, 245, 206)';
export const function_success_fill_solid_03 = 'rgb(149, 229, 153)';
export const function_success_fill_transparent_01 = 'rgba(53, 189, 75, 0.1)';
export const function_success_fill_transparent_02 = 'rgba(53, 189, 75, 0.15)';
export const function_success_fill_transparent_03 = 'rgba(53, 189, 75, 0.2)';
export const function_success_on_success_fill = 'rgb(255, 255, 255)';
export const function_warning_100 = 'rgb(254, 231, 205)';
export const function_warning_200 = 'rgb(254, 196, 139)';
export const function_warning_300 = 'rgb(255, 157, 76)';
export const function_warning_350 = 'rgb(255, 129, 26)';
export const function_warning_400 = 'rgb(237, 109, 12)';
export const function_warning_50 = 'rgb(255, 243, 229)';
export const function_warning_500 = 'rgb(194, 87, 5)';
export const function_warning_600 = 'rgb(164, 73, 4)';
export const function_warning_700 = 'rgb(133, 58, 5)';
export const function_warning_800 = 'rgb(100, 43, 2)';
export const function_warning_900 = 'rgb(59, 26, 2)';
export const function_warning_content_default = 'rgb(237, 109, 12)';
export const function_warning_content_hover = 'rgb(255, 129, 26)';
export const function_warning_content_loading = 'rgb(255, 157, 76)';
export const function_warning_content_pressed = 'rgb(194, 87, 5)';
export const function_warning_fill_default = 'rgb(255, 129, 26)';
export const function_warning_fill_hover = 'rgb(255, 157, 76)';
export const function_warning_fill_loading = 'rgb(255, 157, 76)';
export const function_warning_fill_pressed = 'rgb(237, 109, 12)';
export const function_warning_fill_solid_01 = 'rgb(255, 243, 229)';
export const function_warning_fill_solid_02 = 'rgb(254, 231, 205)';
export const function_warning_fill_solid_03 = 'rgb(254, 196, 139)';
export const function_warning_fill_transparent_01 = 'rgba(237, 109, 12, 0.1)';
export const function_warning_fill_transparent_02 = 'rgba(237, 109, 12, 0.15)';
export const function_warning_fill_transparent_03 = 'rgba(237, 109, 12, 0.2)';
export const function_warning_on_warning_fill = 'rgb(255, 255, 255)';
export const gradient_blue = 'linear-gradient(90deg, rgb(80, 131, 251) 0%, rgb(51, 109, 244) 100%)';
export const gradient_carmine = 'linear-gradient(90deg, rgb(223, 88, 165) 0%, rgb(204, 57, 140) 100%)';
export const gradient_green = 'linear-gradient(90deg, rgb(92, 209, 104) 0%, rgb(53, 189, 75) 100%)';
export const gradient_indigo = 'linear-gradient(90deg, rgb(117, 125, 240) 0%, rgb(91, 101, 245) 100%)';
export const gradient_lime = 'linear-gradient(90deg, rgb(200, 221, 95) 0%, rgb(162, 193, 11) 100%)';
export const gradient_orange = 'linear-gradient(90deg, rgb(255, 157, 76) 0%, rgb(255, 129, 26) 100%)';
export const gradient_purple = 'linear-gradient(90deg, rgb(159, 111, 241) 0%, rgb(141, 85, 237) 100%)';
export const gradient_red = 'linear-gradient(90deg, rgb(255, 117, 112) 0%, rgb(245, 74, 69) 100%)';
export const gradient_turquoise = 'linear-gradient(90deg, rgb(111, 232, 216) 0%, rgb(51, 214, 192) 100%)';
export const gradient_violet = 'linear-gradient(90deg, rgb(222, 129, 222) 0%, rgb(207, 94, 207) 100%)';
export const gradient_wathet = 'linear-gradient(90deg, rgb(62, 195, 247) 0%, rgb(37, 176, 231) 100%)';
export const gradient_yellow = 'linear-gradient(90deg, rgb(255, 198, 10) 0%, rgb(217, 153, 4) 100%)';
export const icon_disabled = 'rgb(187, 191, 196)';
export const icon_n1 = 'rgb(43, 47, 54)';
export const icon_n2 = 'rgb(100, 106, 115)';
export const icon_n3 = 'rgb(143, 149, 158)';
export const illustration_blue_a = 'rgb(0, 34, 112)';
export const illustration_blue_b = 'rgb(0, 34, 112)';
export const illustration_blue_c = 'rgb(0, 34, 112)';
export const illustration_blue_d = 'rgb(0, 34, 112)';
export const illustration_blue_e = 'rgb(51, 109, 244)';
export const illustration_neutral_a = 'rgb(255, 255, 255)';
export const illustration_neutral_b = 'rgb(255, 255, 255)';
export const illustration_neutral_c = 'rgba(187, 191, 196, 0.45)';
export const illustration_neutral_d = 'rgb(143, 149, 158)';
export const illustration_orange = 'rgb(255, 129, 26)';
export const illustration_red = 'rgb(245, 74, 69)';
export const illustration_static_black = 'rgb(0, 0, 0)';
export const illustration_static_white = 'rgb(255, 255, 255)';
export const illustration_turquoise = 'rgb(51, 214, 192)';
export const illustration_wathet = 'rgb(62, 195, 247)';
export const illustration_yellow = 'rgb(255, 198, 10)';
export const inverse_primary_content_default = 'rgb(80, 131, 251)';
export const inverse_primary_content_hover = 'rgb(122, 162, 255)';
export const inverse_primary_content_pressed = 'rgb(51, 109, 244)';
export const line_border_card = 'rgb(222, 224, 227)';
export const line_border_component = 'rgb(208, 211, 214)';
export const line_divider_default = 'rgba(31, 35, 41, 0.15)';
export const line_divider_module = 'rgba(31, 35, 41, 0.15)';
export const primary_content_default = 'rgb(20, 86, 240)';
export const primary_content_hover = 'rgb(51, 109, 244)';
export const primary_content_loading = 'rgb(148, 180, 255)';
export const primary_content_pressed = 'rgb(4, 66, 210)';
export const primary_fill_default = 'rgb(20, 86, 240)';
export const primary_fill_hover = 'rgb(51, 109, 244)';
export const primary_fill_loading = 'rgb(148, 180, 255)';
export const primary_fill_pressed = 'rgb(4, 66, 210)';
export const primary_fill_solid_01 = 'rgb(240, 244, 255)';
export const primary_fill_solid_02 = 'rgb(224, 233, 255)';
export const primary_fill_solid_03 = 'rgb(194, 212, 255)';
export const primary_fill_transparent_01 = 'rgba(20, 86, 240, 0.1)';
export const primary_fill_transparent_02 = 'rgba(20, 86, 240, 0.15)';
export const primary_fill_transparent_03 = 'rgba(20, 86, 240, 0.2)';
export const primary_on_primary_fill = 'rgb(255, 255, 255)';
export const primary_pri_100 = 'rgb(224, 233, 255)';
export const primary_pri_200 = 'rgb(194, 212, 255)';
export const primary_pri_300 = 'rgb(148, 180, 255)';
export const primary_pri_350 = 'rgb(122, 162, 255)';
export const primary_pri_400 = 'rgb(80, 131, 251)';
export const primary_pri_50 = 'rgb(240, 244, 255)';
export const primary_pri_500 = 'rgb(51, 109, 244)';
export const primary_pri_600 = 'rgb(20, 86, 240)';
export const primary_pri_700 = 'rgb(4, 66, 210)';
export const primary_pri_800 = 'rgb(0, 47, 158)';
export const primary_pri_900 = 'rgb(0, 34, 112)';
export const shadow_default = 'rgb(31, 35, 41)';
export const shadow_default_lg = 'rgba(31, 35, 41, 0.08)';
export const shadow_default_md = 'rgba(31, 35, 41, 0.1)';
export const shadow_default_sm = 'rgba(31, 35, 41, 0.12)';
export const shadow_lg_down = '0px 6px 24px 0px rgba(31, 35, 41, 0.08)';
export const shadow_lg_left = '-6px 0px 24px 0px rgba(31, 35, 41, 0.08)';
export const shadow_lg_right = '6px 0px 24px 0px rgba(31, 35, 41, 0.08)';
export const shadow_lg_up = '0px -6px 24px 0px rgba(31, 35, 41, 0.08)';
export const shadow_md_down = '0px 4px 8px 0px rgba(31, 35, 41, 0.1)';
export const shadow_md_left = '-4px 0px 8px 0px rgba(31, 35, 41, 0.1)';
export const shadow_md_right = '4px 0px 8px 0px rgba(31, 35, 41, 0.1)';
export const shadow_md_up = '0px -4px 8px 0px rgba(31, 35, 41, 0.1)';
export const shadow_pri = 'rgb(20, 86, 240)';
export const shadow_pri_lg = 'rgba(20, 86, 240, 0.24)';
export const shadow_pri_lg_down = '0px 6px 24px 0px rgba(20, 86, 240, 0.24)';
export const shadow_pri_md = 'rgba(20, 86, 240, 0.24)';
export const shadow_pri_md_down = '0px 4px 8px 0px rgba(20, 86, 240, 0.24)';
export const shadow_s1_down =
  '0px 1px 2px -2px rgba(31, 35, 41, 0.02), 0px 2px 4px 0px rgba(31, 35, 41, 0.02), 0px 2px 8px 2px rgba(31, 35, 41, 0.02)';
export const shadow_s1_down_pri =
  '0px 1px 2px -2px rgba(20, 86, 240, 0.12), 0px 2px 4px 0px rgba(20, 86, 240, 0.04), 0px 2px 8px 2px rgba(20, 86, 240, 0.02)';
export const shadow_s1_left =
  '-1px 0px 2px -2px rgba(31, 35, 41, 0.02), -2px 0px 4px 0px rgba(31, 35, 41, 0.02), -2px 0px 8px 2px rgba(31, 35, 41, 0.02)';
export const shadow_s1_right =
  '1px 0px 2px -2px rgba(31, 35, 41, 0.02), 2px 0px 4px 0px rgba(31, 35, 41, 0.02), 2px 0px 8px 2px rgba(31, 35, 41, 0.02)';
export const shadow_s1_up =
  '0px -1px 2px -2px rgba(31, 35, 41, 0.02), 0px -2px 4px 0px rgba(31, 35, 41, 0.02), 0px -2px 8px 2px rgba(31, 35, 41, 0.02)';
export const shadow_s2_down =
  '0px 4px 16px 4px rgba(31, 35, 41, 0.03), 0px 4px 8px 0px rgba(31, 35, 41, 0.02), 0px 2px 4px -4px rgba(31, 35, 41, 0.02)';
export const shadow_s2_down_pri =
  '0px 2px 4px -4px rgba(20, 86, 240, 0.12), 0px 4px 8px 0px rgba(20, 86, 240, 0.04), 0px 4px 16px 4px rgba(20, 86, 240, 0.03)';
export const shadow_s2_left =
  '-4px 0px 16px 4px rgba(31, 35, 41, 0.03), -4px 0px 8px 0px rgba(31, 35, 41, 0.02), -2px 0px 4px -4px rgba(31, 35, 41, 0.02)';
export const shadow_s2_right =
  '4px 0px 16px 4px rgba(31, 35, 41, 0.03), 4px 0px 8px 0px rgba(31, 35, 41, 0.02), 2px 0px 4px -4px rgba(31, 35, 41, 0.02)';
export const shadow_s2_up =
  '0px -4px 16px 4px rgba(31, 35, 41, 0.03), 0px -4px 8px 0px rgba(31, 35, 41, 0.02), 0px -2px 4px -4px rgba(31, 35, 41, 0.02)';
export const shadow_s3_down =
  '0px 6px 18px 6px rgba(31, 35, 41, 0.03), 0px 3px 6px -6px rgba(31, 35, 41, 0.05), 0px 4px 8px 0px rgba(31, 35, 41, 0.03)';
export const shadow_s3_down_pri =
  '0px 3px 6px -6px rgba(20, 86, 240, 0.14), 0px 4px 8px 0px rgba(20, 86, 240, 0.06), 0px 6px 18px 6px rgba(20, 86, 240, 0.04)';
export const shadow_s3_left =
  '-6px 0px 18px 6px rgba(31, 35, 41, 0.03), -3px 0px 6px -6px rgba(31, 35, 41, 0.05), -4px 0px 8px 0px rgba(31, 35, 41, 0.03)';
export const shadow_s3_right =
  '6px 0px 18px 6px rgba(31, 35, 41, 0.03), 3px 0px 6px -6px rgba(31, 35, 41, 0.05), 4px 0px 8px 0px rgba(31, 35, 41, 0.03)';
export const shadow_s3_up =
  '0px -6px 18px 6px rgba(31, 35, 41, 0.03), 0px -3px 6px -6px rgba(31, 35, 41, 0.05), 0px -4px 8px 0px rgba(31, 35, 41, 0.03)';
export const shadow_s4_down =
  '0px 8px 24px 8px rgba(31, 35, 41, 0.04), 0px 6px 12px 0px rgba(31, 35, 41, 0.04), 0px 4px 8px -8px rgba(31, 35, 41, 0.06)';
export const shadow_s4_down_pri =
  '0px 4px 8px -4px rgba(20, 86, 240, 0.16), 0px 6px 12px 0px rgba(20, 86, 240, 0.08), 0px 8px 24px 8px rgba(20, 86, 240, 0.06)';
export const shadow_s4_left =
  '-8px 0px 24px 8px rgba(31, 35, 41, 0.04), -6px 0px 12px 0px rgba(31, 35, 41, 0.04), -4px 0px 8px -8px rgba(31, 35, 41, 0.06)';
export const shadow_s4_right =
  '8px 0px 24px 8px rgba(31, 35, 41, 0.04), 6px 0px 12px 0px rgba(31, 35, 41, 0.04), 4px 0px 8px -8px rgba(31, 35, 41, 0.06)';
export const shadow_s4_up =
  '0px -8px 24px 8px rgba(31, 35, 41, 0.04), 0px -6px 12px 0px rgba(31, 35, 41, 0.04), 0px -4px 8px -8px rgba(31, 35, 41, 0.06)';
export const shadow_s5_down =
  '0px 10px 36px 10px rgba(31, 35, 41, 0.04), 0px 8px 24px 0px rgba(31, 35, 41, 0.04), 0px 6px 12px -10px rgba(31, 35, 41, 0.06)';
export const shadow_s5_down_pri =
  '0px 10px 12px -10px rgba(20, 86, 240, 0.18), 0px 8px 24px 0px rgba(20, 86, 240, 0.1), 0px 10px 36px 10px rgba(20, 86, 240, 0.06)';
export const shadow_s5_left =
  '-10px 0px 36px 10px rgba(31, 35, 41, 0.04), -8px 0px 24px 0px rgba(31, 35, 41, 0.04), -6px 0px 12px -10px rgba(31, 35, 41, 0.06)';
export const shadow_s5_right =
  '10px 0px 36px 10px rgba(31, 35, 41, 0.04), 8px 0px 24px 0px rgba(31, 35, 41, 0.04), 6px 0px 12px -10px rgba(31, 35, 41, 0.06)';
export const shadow_s5_up =
  '0px -10px 36px 10px rgba(31, 35, 41, 0.04), 0px -8px 24px 0px rgba(31, 35, 41, 0.04), 0px -6px 12px -10px rgba(31, 35, 41, 0.06)';
export const shadow_sm_down = '0px 2px 4px 0px rgba(31, 35, 41, 0.12)';
export const shadow_sm_left = '-2px 0px 4px 0px rgba(31, 35, 41, 0.12)';
export const shadow_sm_right = '2px 0px 4px 0px rgba(31, 35, 41, 0.12)';
export const shadow_sm_up = '0px -2px 4px 0px rgba(31, 35, 41, 0.12)';
export const static_black = 'rgb(0, 0, 0)';
export const static_black_10 = 'rgba(0, 0, 0, 0.1)';
export const static_black_15 = 'rgba(0, 0, 0, 0.15)';
export const static_black_20 = 'rgba(0, 0, 0, 0.2)';
export const static_black_30 = 'rgba(0, 0, 0, 0.3)';
export const static_black_40 = 'rgba(0, 0, 0, 0.4)';
export const static_black_5 = 'rgba(0, 0, 0, 0.05)';
export const static_black_50 = 'rgba(0, 0, 0, 0.5)';
export const static_black_60 = 'rgba(0, 0, 0, 0.6)';
export const static_black_70 = 'rgba(0, 0, 0, 0.7)';
export const static_black_80 = 'rgba(0, 0, 0, 0.8)';
export const static_black_90 = 'rgba(0, 0, 0, 0.9)';
export const static_white = 'rgb(255, 255, 255)';
export const static_white_10 = 'rgba(255, 255, 255, 0.1)';
export const static_white_15 = 'rgba(255, 255, 255, 0.15)';
export const static_white_20 = 'rgba(255, 255, 255, 0.2)';
export const static_white_30 = 'rgba(255, 255, 255, 0.3)';
export const static_white_40 = 'rgba(255, 255, 255, 0.4)';
export const static_white_5 = 'rgba(255, 255, 255, 0.05)';
export const static_white_50 = 'rgba(255, 255, 255, 0.5)';
export const static_white_60 = 'rgba(255, 255, 255, 0.6)';
export const static_white_70 = 'rgba(255, 255, 255, 0.7)';
export const static_white_80 = 'rgba(255, 255, 255, 0.8)';
export const static_white_90 = 'rgba(255, 255, 255, 0.9)';
export const static_white_hover = 'rgba(255, 255, 255, 0.1)';
export const static_white_pressed = 'rgba(255, 255, 255, 0.2)';
export const text_caption = 'rgb(100, 106, 115)';
export const text_disabled = 'rgb(187, 191, 196)';
export const text_link_disabled = 'rgb(187, 191, 196)';
export const text_link_hover = 'rgb(51, 109, 244)';
export const text_link_loading = 'rgb(148, 180, 255)';
export const text_link_normal = 'rgb(20, 86, 240)';
export const text_link_pressed = 'rgb(4, 66, 210)';
export const text_placeholder = 'rgb(143, 149, 158)';
export const text_title = 'rgb(31, 35, 41)';
