// @universe-design/react/src/transfer/index.tsx
import type { TransferListProps, TransferPanelProps } from './interface';
import { TransferChangeAction } from './interface';
import type { TranferSelectionSummaryProps } from './panel/TargetPanel';
import { TransferTargetPanel, TransferTargetPanelHeader, TranferSelectionSummary } from './panel/TargetPanel';
import type { TransferSelectAllProps } from './panel/SourcePanel';
import { TransferSelectAll, TransferSourcePanel, TransferSourcePanelHeader } from './panel/SourcePanel';
import { TransferSourceList } from './panel/SourceList';
import { TransferSearch } from './panel/Search';
import { TransferTargetList } from './panel/TargetList';
import { Transfer } from './Transfer';
import {
  TransferSourceOption,
  TransferSourceOptionCheckbox,
  TransferSourceOptionLabel,
} from './panel/TransferSourceOption';
import {
  TransferTargetOptionRemoveButton,
  TransferTargetOption,
  TransferTargetOptionLabel,
} from './panel/TransferTargetOption';

export type { TransferOption, TransferGroupOption, TransferRef, TransferProps } from './interface';

export type { TransferContextType } from './hooks/TransferContext';
export { useTransferContext } from './hooks/TransferContext';

export {
  Transfer,
  // Source panel
  TransferSourcePanel,
  TransferSourcePanelHeader,
  TransferSearch,
  TransferSelectAll,
  TransferSourceList,
  // Source option
  TransferSourceOption,
  TransferSourceOptionCheckbox,
  TransferSourceOptionLabel,
  // Target panel
  TransferTargetPanel,
  TransferTargetPanelHeader,
  TranferSelectionSummary,
  TransferTargetList,
  // Target option
  TransferTargetOption,
  TransferTargetOptionLabel,
  TransferTargetOptionRemoveButton,
  // Enum
  TransferChangeAction,
};

export type {
  // source panel
  TransferSelectAllProps,
  // target panel
  TranferSelectionSummaryProps,
  TransferListProps,
  TransferPanelProps,
};
