// @universe-design/react/src/transfer/interface.ts
import type { ListRef } from '@universe-design/rc-virtual-list';
import type * as React from 'react';
import type { TransferPanelContextType } from './hooks/PanelContext';

export type TransferValue = string | number;

export type OptionRenderer = (params: { searchValue: string; isSource: boolean }) => React.ReactNode;
export interface TransferOption<V extends TransferValue = string> {
  /**
   * 下拉菜单选项的值，需要在整个选项列表中保持唯一
   */
  value: V;
  /**
   * 下拉菜单选项显示的标签，如果未指定的话则使用 `value` 作为标签展示
   */
  label?: React.ReactNode;
  /**
   * `label` 的渲染函数方法，可以获取到输入框的 `searchValue`，优先级高于 `label` 属性
   */
  optionRenderer?: OptionRenderer;
  /**
   * 是否为禁用选中
   * @defaultValue false
   */
  disabled?: boolean;
}

export interface TransferGroupOption<
  V extends TransferValue = string,
  Option extends TransferOption<V> = TransferOption<V>,
> {
  /**
   * 分组名称
   */
  title: string;
  /**
   * 该分组下的选项
   */
  children: Option[];
  /**
   * 分组的唯一标识
   */
  key: string;
}

/**
 * onChange 的事件触发来源
 */
export enum TransferChangeAction {
  /**
   * 全选
   */
  SelectAll = 'SelectAll',
  /**
   * 取消全选
   */
  DeSelectAll = 'DeSelectAll',
  /**
   * 选中
   */
  Select = 'Select',
  /**
   * 取消选中或删除某一项
   */
  DeSelect = 'DeSelect',
  /**
   * 点击右侧清空
   */
  ClearAll = 'ClearAll',
}

/**
 * @docgen
 * @public
 * @typeParam V 已选值类型
 * @typeParam Option 选项类型
 */
export type UseTransferProps<V extends TransferValue, Option extends TransferOption<V>> = {
  /**
   * 默认的搜索条件
   */
  defaultSearchValue?: string;
  /**
   * 自定义搜索筛选函数
   */
  filterOption?: (searchValue: string, option: Option) => boolean;
  /**
   * 默认勾选的选项值
   */
  defaultValue?: V[];
  /**
   * 已选选项的受控值
   */
  value?: V[];
  /**
   * onChange
   */
  onChange?: (selectedValues: V[], meta: { action: TransferChangeAction; options: Option[] }) => void;
  /**
   * 选中回调事件
   */
  onSelect?: (value: V, meta: { option: Option }) => void;
  /**
   * 取消选中回调事件
   */
  onDeSelect?: (value: V, meta: { option: Option }) => void;
  /**
   * 当点击清空时的回调
   */
  onClear?: () => void;
  /**
   * 搜索值发生变化时的回调
   */
  onSearch?: (searchValue: string) => void;
  /**
   * 当切换全选状态时
   */
  onToggleSelectAll?: (selectAll: boolean) => void;
} & (
  | {
      group?: false;
      options?: Option[];
    }
  | {
      /**
       * 是否显示分组
       */
      group?: true | [boolean, boolean];
      /**
       * 选项信息
       */
      options?: TransferGroupOption<V, Option>[];
    }
);

export type TransferProps<V extends TransferValue, Option extends TransferOption<V>> = UseTransferProps<V, Option> & {
  /**
   * className 属性透传
   */
  className?: string;
  /**
   * style 属性透传
   */
  style?: React.CSSProperties;
};

export type TransferRef = {
  /**
   * 清除搜索值
   */
  clearSearchValue: () => void;
};

export type TransferPanelProps<
  V extends TransferValue = string,
  OP extends TransferOption<V> = TransferOption<V>,
> = Pick<TransferPanelContextType<V, OP>, 'emptyContent' | 'optionRenderer'> & React.HTMLAttributes<HTMLDivElement>;

/**
 * TransferSourceList 和 TransferTargetList 组件的 props 定义
 */
export type TransferListProps = {
  listRef?: React.Ref<ListRef>;
};
