// @universe-design/react/src/transfer/sx.tsx
import stylex from '@universe-design/stylex';
import { cssTypographyBody0 } from '../stylex';

export const sxTransferLabel = stylex.create({
  paddingLeft: 8,
});

export const sxSelectAll = stylex.create({
  paddingLeft: 4,
  paddingRight: 4,
});

export const sxRemoveButton = stylex.create({
  marginRight: -4,
  color: $token['icon-n2'],
});

export const sxDisabledRemoveButton = stylex.create({
  pointerEvents: 'none',
});

export const sxPanel = stylex.create({
  width: '50% !important',
  paddingTop: '0 !important',
  paddingBottom: '0 !important',
});

export const sxPanelHeader = stylex.create({
  flex: 0,
});

export const sxDisabledTransferCheckbox = stylex.create({ opacity: 0.5 });

export const sxSearch = stylex.create({
  margin: 15,
  marginBottom: 7,
  width: 'calc(100% - 30px)',
  boxSizing: 'border-box',
});

export const sxPanelListWrapper = stylex.create({
  flex: 1,
  minHeight: 0,
  display: 'flex',
  alignItems: 'flex-start',
});

export const sxSelectionSummary = stylex.create({
  paddingTop: 15, // 16 - 1px border
  paddingLeft: 4,
  paddingRight: 4,
  marginBottom: 7,
  position: 'relative',
  boxSizing: 'border-box',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'space-between',
  color: $token['text-title'],
});

export const sxListCaption = stylex.create({
  fontSize: 12,
  fontWeight: 400,
  color: $token['text-caption'],
});

export const sxClearButton = stylex.create({
  marginRight: 8,
});

export const sxPanelEmpty = [
  stylex.create({
    textAlign: 'center',
    margin: 'auto',
    color: $token['text-placeholder'],
  }),
  cssTypographyBody0,
];

export const sxTransfer = stylex.create({
  height: 320,
  boxShadow: 'none',
  padding: 0,
  margin: 0,
});

export const sxOption = stylex.create({
  marginLeft: 7,
  marginRight: 7,
});

export const sxPanelListAutoSizer = stylex.create({
  width: '100%',
});

export const sxPanelListScrolling = stylex.create({
  '::before': {
    display: 'block',
    content: '""',
    boxShadow: `0 0 0 .5px ${$token['line-divider-default']}`,
  },
});

export const sxLabelHighlight = stylex.create({
  color: $token['primary-content-default'],
});
