// @universe-design/react/src/metrics/MetricsSecondaryContent.tsx
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx, makeStylexStyles } from '@universe-design/stylex/react';
import ExpandDownFilled from '@universe-design/icons-react/ExpandDownFilled';
import ExpandUpFilled from '@universe-design/icons-react/ExpandUpFilled';
import stylex from '@universe-design/stylex';
import React, { useCallback, useState } from 'react';
import { Divider } from '../divider';
import { Text } from '../text';
import { type MetricsProps, type SecondaryInfo } from './Metrics';
import type { ElementOf } from '@universe-design/react-utilities';
import { ResizeObserver, useMergedRefs } from '@universe-design/react-utilities';
import { forwardHTMLRef } from '../util/forwardRef';
import { useMetricsContext } from './MetricsContext';
import { presetColor } from './constant';
import { useMetricsSecondaryDescendant } from './useMetrics';
import { useMetricsGroupContextSelector } from './MetricsGroup';
import clsx from 'clsx';

function isPresetColor(color: string): color is ElementOf<typeof presetColor> {
  return presetColor.includes(color as ElementOf<typeof presetColor>);
}

type MetricsSecondaryContentProps = Pick<MetricsProps, 'secondary'>;

const SecondaryItem = ({
  index,
  setItemWidths,
  item,
}: {
  index: number;
  item: SecondaryInfo;
  setItemWidths: React.Dispatch<React.SetStateAction<number[]>>;
}) => (
  <ResizeObserver
    onResize={({ width }) => {
      setItemWidths(prev => {
        const newWidths = [...prev];
        newWidths[index] = width;
        return newWidths;
      });
    }}
  >
    <MetricsSecondaryContentItem
      ref={useCallback(
        (node: HTMLSpanElement | null) => {
          if (!node) {
            setItemWidths(prev => {
              const newWidths = [...prev];
              newWidths.splice(index, 1);
              return newWidths;
            });
          }
        },
        [index, setItemWidths],
      )}
      {...item}
    />
  </ResizeObserver>
);

export const MetricsSecondaryContent = forwardHTMLRef<MetricsSecondaryContentProps, 'div'>((props, ref) => {
  const { secondary, className, ...restProps } = props;
  const secondaryInfo = secondary?.data;
  const [itemWidths, setItemWidths] = useState<number[]>([]);
  const [containerWidth, setContainerWidth] = useState(0);
  const [dividerWidth, setDividerWidth] = useState(0);

  const { index, register } = useMetricsSecondaryDescendant();
  const recalculateSecondaryHeights = useMetricsGroupContextSelector(context => context?.recalculateSecondaryHeights);
  const secondaryRowMaxHeight = useMetricsGroupContextSelector(
    context => context?.secondaryRowInfo?.[index]?.maxHeight,
  );

  const { rootCssScope } = useMetricsContext();
  const styles = useMetricsSecondaryItemStylexStyles({
    prefix: rootCssScope(),
  });

  interface LayoutItem {
    item: SecondaryInfo;
    isLast: boolean;
    marginRight: number;
    width: number;
  }

  const layoutSingleRow = React.useCallback(
    (items: SecondaryInfo[], dividerWidth: number, startIndex: number): LayoutItem[] => {
      const row: LayoutItem[] = [];
      let totalWidth = itemWidths[startIndex];

      // 添加第一个元素
      row.push({
        item: items[startIndex],
        isLast: false,
        marginRight: 0,
        width: itemWidths[startIndex],
      });

      let currentIndex = startIndex + 1;

      // 尝试添加更多元素
      while (currentIndex < items.length) {
        const nextItemWidth = dividerWidth + itemWidths[currentIndex];
        const newTotalWidth = totalWidth + nextItemWidth;

        if (newTotalWidth > containerWidth) {
          // 当前行放不下更多元素
          const lastItem = row[row.length - 1];
          lastItem.isLast = true;
          lastItem.marginRight = Math.floor(containerWidth - totalWidth) || 0;
          break;
        }

        // 可以继续添加元素
        row.push({
          item: items[currentIndex],
          isLast: false,
          marginRight: 0,
          width: itemWidths[currentIndex],
        });

        totalWidth = newTotalWidth;
        currentIndex++;
      }

      // 处理最后一行
      if (currentIndex === items.length) {
        const lastItem = row[row.length - 1];
        lastItem.isLast = true;
        lastItem.marginRight = Math.floor(containerWidth - totalWidth) || 0;
      }

      return row;
    },
    [containerWidth, itemWidths],
  );

  const secondaryList = React.useMemo(() => {
    if (!secondaryInfo) {
      return [];
    }

    const result: LayoutItem[] = [];
    let currentIndex = 0;

    while (currentIndex < secondaryInfo.length) {
      const row = layoutSingleRow(secondaryInfo, dividerWidth, currentIndex);

      result.push(...row);
      currentIndex += row.length;
    }

    return result;
  }, [secondaryInfo, layoutSingleRow, dividerWidth]);

  if (!secondaryInfo) {
    return null;
  }

  return (
    <React.Fragment>
      <div
        className={clsx(`${rootCssScope()}__secondary__wrapper`, className)}
        style={{
          height: secondaryRowMaxHeight,
        }}
        {...restProps}
      >
        <ResizeObserver
          onResize={({ width }) => {
            setContainerWidth(width);
            recalculateSecondaryHeights?.();
          }}
        >
          <div ref={useMergedRefs(ref, register)}>
            {secondaryList.map(({ item, isLast, marginRight }, i) => (
              <React.Fragment key={i}>
                <SecondaryItem index={i} item={item} setItemWidths={setItemWidths} />
                {isLast && marginRight > 0 && <span style={{ width: marginRight }} sx={[styles.spacer]} />}
                {!isLast && (
                  <ResizeObserver
                    onResize={({ width }) => {
                      setDividerWidth(width);
                    }}
                  >
                    <span
                      style={{
                        display: 'inline-block',
                      }}
                    >
                      <Divider type="vertical" />
                    </span>
                  </ResizeObserver>
                )}
              </React.Fragment>
            ))}
          </div>
        </ResizeObserver>
      </div>
    </React.Fragment>
  );
});

const useMetricsSecondaryItemStylexStyles = makeStylexStyles(({ prefix }: { prefix: string }) => ({
  root: stylex.create(
    {
      display: 'inline-block',
      whiteSpace: 'nowrap',
    },
    `${prefix}__root`,
  ),
  spacer: stylex.create(
    {
      display: 'inline-block',
    },
    `${prefix}__spacer`,
  ),
  label: stylex.create(
    {
      color: $token['text-caption'],
      marginRight: 4,
    },
    `${prefix}__label`,
  ),
  icon: stylex.create(
    {
      fontFamily: 'Roboto',
    },
    `${prefix}__icon`,
  ),
  value: stylex.create(
    {
      marginLeft: 2,
      fontFamily: 'Roboto',
      lineHeight: '100%',
    },
    `${prefix}__value`,
  ),
  success: stylex.create(
    {
      color: $token['datav-function-element-success'],
    },
    `${prefix}__success`,
  ),
  danger: stylex.create(
    {
      color: $token['datav-function-element-danger'],
    },
    `${prefix}__danger`,
  ),
  info: stylex.create(
    {
      color: $token['datav-function-element-info'],
    },
    `${prefix}__info`,
  ),
  warning: stylex.create(
    {
      color: $token['datav-function-element-warning'],
    },
    `${prefix}__warning`,
  ),
  neutral: stylex.create(
    {
      color: $token['datav-function-element-neutral'],
    },
    `${prefix}__neutral`,
  ),
}));

type MetricsSecondaryContentItemProps = Pick<SecondaryInfo, 'label' | 'value' | 'type' | 'valueStyle' | 'color'>;
const MetricsSecondaryContentItem = forwardHTMLRef<MetricsSecondaryContentItemProps, 'span'>((props, ref) => {
  const { label, value: valueNode, type = 'none', valueStyle, color = 'neutral', ...restProps } = props;
  const { size, rootCssScope } = useMetricsContext();

  const iconMap = {
    increase: <ExpandUpFilled />,
    decrease: <ExpandDownFilled />,
    none: null,
  };
  const iconNode = iconMap[type];
  const isPreset = isPresetColor(color);

  const styles = useMetricsSecondaryItemStylexStyles({
    prefix: rootCssScope(),
  });

  return (
    <span ref={ref} sx={[styles.root]} {...restProps}>
      {label !== null && (
        <Text variant="body0" sx={[styles.label]}>
          {label}
        </Text>
      )}
      {iconNode && (
        <Text
          sx={[styles.icon, isPreset ? styles[color] : null]}
          style={{
            color: isPreset ? undefined : color,
            ...valueStyle,
          }}
          variant="headline"
        >
          {iconNode}
        </Text>
      )}
      {valueNode !== null && (
        <Text
          sx={[styles.value, isPreset ? styles[color] : null]}
          style={{
            color: isPreset ? undefined : color,
            ...valueStyle,
          }}
          variant={size === 'medium' ? 'headline' : 'title4'}
        >
          {valueNode}
        </Text>
      )}
    </span>
  );
});
