// @universe-design/react/src/metrics/MetricsMainContent.tsx
/* eslint-disable react/no-unknown-property */
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx, makeStylexStyles } from '@universe-design/stylex/react';
import { ResizeObserver, useMergedRefs } from '@universe-design/react-utilities';
import stylex from '@universe-design/stylex';
import React, { useMemo } from 'react';
import { type MetricsProps } from './Metrics';
import { MetricsChart } from './MetricsChart';
import { forwardHTMLRef } from '../util/forwardRef';
import { useMetricsContext } from './MetricsContext';
import { resolveShorthandOld } from '../shared/slot';
import { useMetricsMainDescendant } from './useMetrics';
import { useMetricsGroupContextSelector } from './MetricsGroup';

type MetricsMainContentProps = Pick<MetricsProps, 'mainValue' | 'chart'>;

const useMetricsMainContentStylexStyles = makeStylexStyles(({ prefix }: { prefix: string }) => ({
  root: stylex.create(
    {
      width: '100%',
      display: 'flex',
      alignItems: 'center',
      flexWrap: 'nowrap',
      justifyContent: 'space-between',
      fontFamily: 'Roboto',
      color: $token['text-title'],
      fontWeight: 500,
    },
    `${prefix}__root`,
  ),
  mainWrapper: stylex.create(
    {
      display: 'flex',
      alignItems: 'center',
    },
    `${prefix}__main__wrapper`,
  ),
  medium: stylex.create(
    {
      fontSize: 24,
      fontWeight: 500,
      lineHeight: '28px',
    },
    `${prefix}__medium`,
  ),
  large: stylex.create(
    {
      fontSize: 36,
      lineHeight: '42px',
    },
    `${prefix}__large`,
  ),
  container: stylex.create(
    {
      width: '100%',
      position: 'relative',
    },
    `${prefix}__container`,
  ),
  content: stylex.create(
    {
      position: 'absolute',
      transformOrigin: 'left top',
      whiteSpace: 'nowrap',
    },
    `${prefix}__content`,
  ),
}));

export const MetricsMainContent = forwardHTMLRef<MetricsMainContentProps, 'div'>((props, ref) => {
  const { mainValue, chart, ...restProps } = props;

  const { rootCssScope, size, position } = useMetricsContext();

  const { index, register } = useMetricsMainDescendant();
  const recalculateMainHeights = useMetricsGroupContextSelector(context => context?.recalculateMainHeights);
  const mainRowMaxHeight = useMetricsGroupContextSelector(context => context?.mainRowInfo?.[index]?.maxHeight);

  const styles = useMetricsMainContentStylexStyles({
    prefix: rootCssScope(),
  });

  const [containerWidth, setContainerWidth] = React.useState(1);
  const [contentWidth, setContentWidth] = React.useState(1);
  const [contentHeight, setContentHeight] = React.useState(1);
  const scale = useMemo(() => {
    if (containerWidth < contentWidth) {
      return containerWidth / contentWidth;
    }
    return 1;
  }, [containerWidth, contentWidth]);

  const mainValueProps = resolveShorthandOld(mainValue);
  const { autoScaleContent = true } = mainValueProps;

  let content: React.ReactNode;

  if (!autoScaleContent) {
    content = <span>{mainValueProps?.children}</span>;
  } else {
    content = (
      <ResizeObserver
        onResize={({ width, height }) => {
          setContentWidth(width);
          setContentHeight(height);
        }}
      >
        <span
          sx={[styles.content]}
          style={{
            transform: `scale(${scale})`,
          }}
        >
          {mainValueProps?.children}
        </span>
      </ResizeObserver>
    );
  }

  return (
    <div
      sx={[styles.mainWrapper]}
      style={{
        height: mainRowMaxHeight,
      }}
      {...restProps}
    >
      <ResizeObserver onResize={recalculateMainHeights}>
        <div ref={useMergedRefs(ref, register)} sx={[styles.root, styles[size]]}>
          <ResizeObserver
            onResize={({ width }) => {
              setContainerWidth(width);
            }}
          >
            <div
              sx={[styles.container]}
              style={
                autoScaleContent
                  ? {
                      height: contentHeight * scale,
                    }
                  : {}
              }
            >
              {content}
            </div>
          </ResizeObserver>
          {position === 'afterMainValue' && <MetricsChart chart={chart} />}
        </div>
      </ResizeObserver>
    </div>
  );
});
