// @universe-design/react/src/metrics/Metrics.tsx
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx } from '@universe-design/stylex/react';
import * as React from 'react';
import { type DefaultSizeType, type ElementOf, type LiteralUnion } from '@universe-design/react-utilities';
import { forwardHTMLRef } from '../util/forwardRef';
import type { TooltipProps } from '../tooltip';
import { MetricsProvider } from './MetricsContext';
import type { presetColor } from './constant';
import { resolveShorthandOld, type WithSlotShorthandValue } from '../shared/slot';
import { MetricsRoot } from './MetricsRoot';
import './style';

export interface ChartPosition {
  /**
   * 图表区域（整体右侧、主数据右侧）
   * @defaultValue `afterAll`
   */
  position?: 'afterAll' | 'afterMainValue';
}

export interface MetricsProps {
  /**
   * Metrics 尺寸
   * @defaultValue `medium`
   */
  size?: Extract<DefaultSizeType, 'medium' | 'large'>;
  /**
   * 指标标题
   */
  mainTitle?: WithSlotShorthandValue<React.ComponentProps<'div'>> | null;
  /**
   * info 内容
   */
  mainInfoIcon?: WithSlotShorthandValue<{
    tooltip?: TooltipProps;
  }> | null;
  /**
   * info 右侧自定义内容
   */
  mainHeaderContentAfter?: WithSlotShorthandValue<React.ComponentProps<'div'>> | null;
  /**
   * 数值内容
   */
  mainValue: WithSlotShorthandValue<
    React.ComponentProps<'div'> & {
      /**
       * 自动根据容器和内容缩放文字区域
       */
      autoScaleContent?: boolean;
    }
  >;
  /**
   * 次要指标
   */
  secondary?: WithSlotShorthandValue<{
    data: SecondaryInfo[];
  }>;
  /**
   * 次要内容下面自定义区域
   */
  footer?: WithSlotShorthandValue<React.ComponentProps<'div'>> | null;
  /**
   * 图表
   */
  chart?: WithSlotShorthandValue<React.ComponentProps<'div'> & ChartPosition> | null;
}

export interface SecondaryInfo {
  /**
   * 次要指标趋势类型（增、减、无）
   * @defaultValue `none`
   */
  type?: 'increase' | 'decrease' | 'none';
  /**
   * 次要指标颜色，包含预设状态颜色
   * @defaultValue `neutral`
   */
  color?: LiteralUnion<ElementOf<typeof presetColor>, string>;
  /**
   * 次要指标标题
   */
  label: React.ReactNode;
  /**
   * 次级数值
   */
  value: React.ReactNode;
  /**
   * 数值样式
   */
  valueStyle?: React.CSSProperties;
}

export const Metrics = forwardHTMLRef<MetricsProps, 'div'>((props, ref) => {
  const chartProps = resolveShorthandOld(props.chart);
  const { position } = chartProps || {};

  return (
    <MetricsProvider size={props.size} position={position}>
      <MetricsRoot ref={ref} {...props} />
    </MetricsProvider>
  );
});
