// @universe-design/react/src/metrics/MetricsRoot.tsx
/* eslint-disable react-hooks/rules-of-hooks */
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx, makeStylexStyles } from '@universe-design/stylex/react';
import { MetricsHeader } from './MetricsHeader';
import { MetricsMainContent } from './MetricsMainContent';
import { MetricsChart } from './MetricsChart';
import { MetricsSecondaryContent } from './MetricsSecondaryContent';
import stylex from '@universe-design/stylex';
import { useMetricsContext } from './MetricsContext';
import { forwardHTMLRef } from '../internal';
import * as React from 'react';
import type { MetricsProps } from './Metrics';
import { useMergedRefs, ResizeObserver } from '@universe-design/react-utilities';
import { useMetricsDescendant } from './useMetrics';
import { useMetricsGroupContextSelector } from './MetricsGroup';
import { MetricsFooter } from './MetricsFooter';

interface MetricsValueProps {
  children?: React.ReactNode;
}

const sxMetricsValue = {
  root: stylex.create({
    flex: 1,
    width: '100%',
  }),
};
const MetricsValue = forwardHTMLRef<MetricsValueProps, 'div'>((props, ref) => (
  <div ref={ref} className={stylex(sxMetricsValue.root)}>
    {props.children}
  </div>
));

const useMetricsStylexStyles = makeStylexStyles(({ prefix }: { prefix: string }) => ({
  root: stylex.create(
    {
      borderRadius: $token['border-radius-lg'],
      display: 'flex',
      alignItems: 'center',
      gap: 8,
    },
    `${prefix}__root`,
  ),
  main: stylex.create(
    {
      gap: 4,
      display: 'flex',
      flexDirection: 'column',
    },
    `${prefix}__main`,
  ),
  'secondary-root': stylex.create(
    {
      display: 'flex',
      flexDirection: 'column',
    },
    `${prefix}__secondary-root`,
  ),
  'secondary-medium': stylex.create(
    {
      gap: 8,
    },
    `${prefix}__secondary-medium`,
  ),
  'secondary-large': stylex.create(
    {
      gap: 12,
    },
    `${prefix}__secondary-large`,
  ),
}));

export const MetricsRoot = React.memo(
  forwardHTMLRef<MetricsProps, 'div'>((props, ref) => {
    const { mainTitle, mainInfoIcon, mainHeaderContentAfter, mainValue, secondary, footer, chart, ...restProps } =
      props;

    const { register } = useMetricsDescendant();

    const { rootCssScope, size, position } = useMetricsContext();
    const recalculateLocation = useMetricsGroupContextSelector(context => context?.recalculateLocation);

    const styles = useMetricsStylexStyles({
      prefix: rootCssScope(),
    });

    return (
      <ResizeObserver onResize={recalculateLocation}>
        <div ref={useMergedRefs(ref, register)} sx={[styles.root]} {...restProps}>
          <MetricsValue>
            <div sx={[styles[`secondary-${size}`], styles['secondary-root']]}>
              <div sx={[styles.main]}>
                <MetricsHeader
                  mainTitle={mainTitle}
                  mainInfoIcon={mainInfoIcon}
                  mainHeaderContentAfter={mainHeaderContentAfter}
                />
                <MetricsMainContent mainValue={mainValue} chart={chart} />
              </div>
              <MetricsSecondaryContent secondary={secondary} />
            </div>
            <MetricsFooter footer={footer} />
          </MetricsValue>
          {position === 'afterAll' && <MetricsChart chart={chart} />}
        </div>
      </ResizeObserver>
    );
  }),
);
