// @universe-design/react/src/metrics/MetricsHeader.tsx
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx, makeStylexStyles } from '@universe-design/stylex/react';
import InfoOutlined from '@universe-design/icons-react/InfoOutlined';
import stylex from '@universe-design/stylex';
import Tooltip from '../tooltip';
import { Text } from '../text';
import { type MetricsProps } from './Metrics';
import { forwardHTMLRef } from '../util/forwardRef';
import { useMetricsContext } from './MetricsContext';
import { resolveShorthandOld } from '../shared/slot';
import { useMetricsGroupContextSelector } from './MetricsGroup';
import { useMetricsHeaderDescendant } from './useMetrics';
import { useMergedRefs, ResizeObserver } from '@universe-design/react-utilities';
import * as React from 'react';
import clsx from 'clsx';
import './style';

type MetricsHeaderProps = Pick<MetricsProps, 'mainTitle' | 'mainInfoIcon' | 'mainHeaderContentAfter'>;

const useMetricsHeaderStylexStyles = makeStylexStyles(({ prefix }: { prefix: string }) => ({
  header: stylex.create(
    {
      width: '100%',
      display: 'flex',
      alignItems: 'baseline',
      gap: 4,
    },
    `${prefix}__header`,
  ),
  icon: stylex.create(
    {
      color: $token['icon-n2'],
      fontSize: 'calc(1em + 2px)',
      verticalAlign: 'calc(0.36 * 1em - 0.5 * (1em + 2px))',
    },
    `${prefix}__icon`,
  ),
}));

export const MetricsHeader = forwardHTMLRef<MetricsHeaderProps, 'div'>((props, ref) => {
  const { mainTitle, mainInfoIcon, mainHeaderContentAfter, className, ...restProps } = props;

  const { rootCssScope, size } = useMetricsContext();
  const { index, register } = useMetricsHeaderDescendant();
  const recalculateHeaderHeights = useMetricsGroupContextSelector(context => context?.recalculateHeaderHeights);
  const headerRowMaxHeight = useMetricsGroupContextSelector(context => context?.headerRowInfo?.[index]?.maxHeight);
  const titleLineClamp = useMetricsGroupContextSelector(context => context?.titleLineClamp) || 2;

  const styles = useMetricsHeaderStylexStyles({
    prefix: rootCssScope(),
  });

  const mainTitleProps = resolveShorthandOld(mainTitle);
  const mainInfoIconProps = resolveShorthandOld(mainInfoIcon);
  const mainHeaderContentAfterProps = resolveShorthandOld(mainHeaderContentAfter);

  const { tooltip } = mainInfoIconProps || {};

  return (
    <React.Fragment>
      <div
        className={clsx(`${rootCssScope()}__header__wrapper`, className)}
        style={{
          height: headerRowMaxHeight,
        }}
        {...restProps}
      >
        <ResizeObserver onResize={recalculateHeaderHeights}>
          <Text
            as="div"
            ref={useMergedRefs(ref, register)}
            sx={[styles.header]}
            variant={size === 'medium' ? 'body0' : 'title5'}
          >
            {mainTitle !== null && <Text lineClamp={titleLineClamp}>{mainTitleProps?.children}</Text>}
            {mainInfoIcon !== null && (
              <Tooltip {...tooltip!}>
                <span>
                  <InfoOutlined className={stylex(styles.icon)} width={1} />
                </span>
              </Tooltip>
            )}
            {mainHeaderContentAfter !== null && <div>{mainHeaderContentAfterProps?.children}</div>}
          </Text>
        </ResizeObserver>
      </div>
    </React.Fragment>
  );
});
