// @universe-design/react/src/metrics/MetricsGroup.tsx
import { createContextHook } from '@universe-design/react-provider';
import { createContext, useContext, useContextSelector } from 'use-context-selector';
import * as React from 'react';
import {
  MetricsDescendantsProvider,
  MetricsHeaderDescendantsProvider,
  MetricsSecondaryDescendantsProvider,
  useMetricsDescendants,
  useMetricsFooterDescendants,
  useMetricsHeaderDescendants,
  useMetricsSecondaryDescendants,
  MetricsFooterDescendantsProvider,
  useMetricsMainDescendants,
  MetricsMainDescendantsProvider,
} from './useMetrics';
import { ResizeObserver } from '@universe-design/react-utilities';
import { forwardHTMLRef } from '../internal';
import { useCssScope } from '../config-provider';
import clsx from 'clsx';
import { useSyncExternalStore } from 'use-sync-external-store/shim';

interface RowInfo {
  row: number; // 当前所在行
  maxHeight: number; // 当前区域最大高度
}

interface MetricsGroupContextType extends MetricsGroupProps {
  /**
   * 重新计算 左上角坐标
   */
  recalculateLocation: () => void;
  /**
   * 重新计算 header 高度
   */
  recalculateHeaderHeights: () => void;
  /**
   * 重新计算 主指标 高度
   */
  recalculateMainHeights: () => void;
  /**
   * 重新计算 次级指标 高度
   */
  recalculateSecondaryHeights: () => void;
  /**
   * 重新计算 footer 高度
   */
  recalculateFooterHeights: () => void;
  /**
   * 主指标 高度信息
   */
  mainRowInfo: RowInfo[];
  /**
   * header 高度信息
   */
  headerRowInfo: RowInfo[];
  /**
   * 次级指标 高度信息
   */
  secondaryRowInfo: RowInfo[];
  /**
   * footer 高度信息
   */
  footerRowInfo: RowInfo[];
}

const { useContextSelector: useMetricsGroupContextSelector, Provider: MetricsGroupContextProvider } =
  createContextHook<MetricsGroupContextType | null>('MetricsGroupContext', {
    strict: false,
    customizedContextMethods: {
      createContext,
      useContext,
      useContextSelector,
    },
    defaultValue: null,
  });

export interface MetricsGroupProps {
  /**
   * 标题行数，默认双行省略
   * @defaultValue 2
   */
  titleLineClamp?: number;
}

const MetricsGroup = forwardHTMLRef<MetricsGroupProps, 'div'>((props, ref) => {
  const { titleLineClamp = 2, className, ...restProps } = props;
  const [location, setLocation] = React.useState<{ x: number; y: number }[]>([]);
  const [headerHeights, setHeaderHeights] = React.useState<number[]>([]);
  const [mainHeights, setMainHeights] = React.useState<number[]>([]);
  const [secondaryHeights, setSecondaryHeights] = React.useState<number[]>([]);
  const [footerHeights, setFooterHeights] = React.useState<number[]>([]);
  const rootCssScope = useCssScope('metrics');

  const descendants = useMetricsDescendants();
  const { list: descendantList } = useSyncExternalStore(
    descendants.subscribe,
    descendants.getSnapshot,
    descendants.getSnapshot,
  );

  const headerDescendants = useMetricsHeaderDescendants();
  const { list: headerDescendantList } = useSyncExternalStore(
    headerDescendants.subscribe,
    headerDescendants.getSnapshot,
    headerDescendants.getSnapshot,
  );
  const mainDescendants = useMetricsMainDescendants();
  const { list: mainDescendantList } = useSyncExternalStore(
    mainDescendants.subscribe,
    mainDescendants.getSnapshot,
    mainDescendants.getSnapshot,
  );
  const secondaryDescendants = useMetricsSecondaryDescendants();
  const { list: secondaryDescendantList } = useSyncExternalStore(
    secondaryDescendants.subscribe,
    secondaryDescendants.getSnapshot,
    secondaryDescendants.getSnapshot,
  );
  const footerDescendants = useMetricsFooterDescendants();
  const { list: footerDescendantList } = useSyncExternalStore(
    footerDescendants.subscribe,
    footerDescendants.getSnapshot,
    footerDescendants.getSnapshot,
  );

  const recalculateLocation = React.useCallback(() => {
    const xy: { x: number; y: number }[] = [];
    for (const d of descendantList) {
      xy.push({
        x: d?.node.offsetLeft || 0,
        y: d?.node.offsetTop || 0,
      });
    }

    setLocation(xy);
  }, [descendantList]);

  const recalculateHeaderHeights = React.useCallback(() => {
    const headerHeight: number[] = [];
    for (const d of headerDescendantList) {
      headerHeight.push(d?.node.clientHeight as number);
    }
    setHeaderHeights(headerHeight);
  }, [headerDescendantList]);

  const recalculateMainHeights = React.useCallback(() => {
    const mainHeight: number[] = [];
    for (const d of mainDescendantList) {
      mainHeight.push(d?.node.clientHeight as number);
    }
    setMainHeights(mainHeight);
  }, [mainDescendantList]);

  const recalculateSecondaryHeights = React.useCallback(() => {
    const secondaryHeight: number[] = [];
    for (const d of secondaryDescendantList) {
      secondaryHeight.push(d?.node.clientHeight as number);
    }
    setSecondaryHeights(secondaryHeight);
  }, [secondaryDescendantList]);

  const recalculateFooterHeights = React.useCallback(() => {
    const footerHeight: number[] = [];
    for (const d of footerDescendantList) {
      footerHeight.push(d?.node.clientHeight as number);
    }
    setFooterHeights(footerHeight);
  }, [footerDescendantList]);

  React.useEffect(() => {
    recalculateLocation();
    recalculateHeaderHeights();
    recalculateMainHeights();
    recalculateSecondaryHeights();
    recalculateFooterHeights();
  }, [
    recalculateLocation,
    recalculateHeaderHeights,
    recalculateMainHeights,
    recalculateSecondaryHeights,
    recalculateFooterHeights,
  ]);

  const calculateRowInfo = React.useCallback(
    (heights: number[]) => {
      const rows: { y: number; heights: number[] }[] = [];

      // 按 y 坐标分组
      location.forEach((loc, index) => {
        const existingRow = rows.find(row => row.y === loc.y);
        if (existingRow) {
          existingRow.heights.push(heights[index]);
        } else {
          rows.push({ y: loc.y, heights: [heights[index]] });
        }
      });

      // 计算每行的最大高度映射
      const rowMaxHeights = rows.map(row => ({
        y: row.y,
        maxHeight: Math.max(...row.heights),
      }));

      // 为每个位置生成对应的行信息
      return location.map(loc => ({
        row: rows.findIndex(r => r.y === loc.y),
        maxHeight: rowMaxHeights.find(r => r.y === loc.y)?.maxHeight || 0,
      }));
    },
    [location],
  );

  const headerRowInfo = React.useMemo(() => calculateRowInfo(headerHeights), [calculateRowInfo, headerHeights]);

  const mainRowInfo = React.useMemo(() => calculateRowInfo(mainHeights), [calculateRowInfo, mainHeights]);

  const secondaryRowInfo = React.useMemo(
    () => calculateRowInfo(secondaryHeights),
    [calculateRowInfo, secondaryHeights],
  );

  const footerRowInfo = React.useMemo(() => calculateRowInfo(footerHeights), [calculateRowInfo, footerHeights]);

  return (
    <MetricsDescendantsProvider value={descendants}>
      <MetricsHeaderDescendantsProvider value={headerDescendants}>
        <MetricsSecondaryDescendantsProvider value={secondaryDescendants}>
          <MetricsFooterDescendantsProvider value={footerDescendants}>
            <MetricsMainDescendantsProvider value={mainDescendants}>
              <MetricsGroupContextProvider
                value={{
                  titleLineClamp,
                  recalculateLocation,
                  recalculateMainHeights,
                  recalculateHeaderHeights,
                  recalculateSecondaryHeights,
                  recalculateFooterHeights,
                  headerRowInfo,
                  mainRowInfo,
                  secondaryRowInfo,
                  footerRowInfo,
                }}
              >
                <ResizeObserver onResize={recalculateLocation}>
                  <div className={clsx(`${rootCssScope()}__group__wrapper`, className)} ref={ref} {...restProps}>
                    {props.children}
                  </div>
                </ResizeObserver>
              </MetricsGroupContextProvider>
            </MetricsMainDescendantsProvider>
          </MetricsFooterDescendantsProvider>
        </MetricsSecondaryDescendantsProvider>
      </MetricsHeaderDescendantsProvider>
    </MetricsDescendantsProvider>
  );
});

export { useMetricsGroupContextSelector, MetricsGroup };
