// @universe-design/react/lib/image-viewer/components/Download.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useImageViewerToolbarDownload = exports.Download = void 0;
var React = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _DownloadOutlined = _interopRequireDefault(require("@universe-design/icons-react/DownloadOutlined"));
var _hooks = require("../../config-provider/hooks");
var _ImageViewerContext = require("../ImageViewerContext");
var _tooltip = _interopRequireDefault(require("../../tooltip"));
var _button = require("../../button");
require("../style");
function downloadByLink(blobUrl, name) {
  const a = document.createElement('a');
  a.style.display = 'none';
  a.href = blobUrl;
  a.download = name;
  document.body.appendChild(a);
  a.click();
  document.body.removeChild(a);
}
const downloadBlob = async (blob, name) => {
  const blobUrl = window.URL.createObjectURL(blob);
  downloadByLink(blobUrl, name);
  window.URL.revokeObjectURL(blobUrl);
};
const downloadImage = async (source, name) => {
  if (source instanceof Blob) {
    await downloadBlob(source, name);
    return;
  }
  if (source.startsWith('blob:')) {
    downloadByLink(source, name);
    return;
  }
  const blob = await fetch(source).then(r => r.blob());
  await downloadBlob(blob, name);
};

/**
 * 图片查看器下载工具 hook
 *
 * @public
 */
const useImageViewerToolbarDownload = () => {
  const {
    images,
    currentIndex
  } = (0, _ImageViewerContext.useImageViewerContext)();
  const download = React.useCallback(async () => {
    var _images$currentIndex$;
    await downloadImage(images[currentIndex].src, (_images$currentIndex$ = images[currentIndex].alt) !== null && _images$currentIndex$ !== void 0 ? _images$currentIndex$ : `${new Date().getTime()}.jpg`);
  }, [currentIndex, images]);
  return {
    download
  };
};
exports.useImageViewerToolbarDownload = useImageViewerToolbarDownload;
const Download = exports.Download = /*#__PURE__*/React.memo(_ref => {
  let {
    onDownload
  } = _ref;
  const locale = (0, _hooks.useLocaleReceiver)('imageViewer');
  const {
    rootCssScope
  } = (0, _ImageViewerContext.useImageViewerContext)();
  const {
    images,
    currentIndex
  } = (0, _ImageViewerContext.useImageViewerContext)();
  const {
    download
  } = useImageViewerToolbarDownload();
  return /*#__PURE__*/React.createElement(_tooltip.default, {
    title: locale.download,
    overlayClassName: rootCssScope('__toolbar__tooltip')
  }, /*#__PURE__*/React.createElement(_button.IconButton, {
    size: 20,
    icon: /*#__PURE__*/React.createElement(_DownloadOutlined.default, null),
    className: (0, _clsx.default)(rootCssScope('__toolbar__button'), rootCssScope('__toolbar__download')),
    onClick: onDownload ? () => onDownload(images[currentIndex], {
      downloadImage
    }) : download
  }));
});
