// @universe-design/react/lib/image-viewer/components/Pagination.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useImageViewerToolbarPagination = exports.Pagination = void 0;
var _ukchsq = _interopRequireDefault(require("./../../.stylex/ukchsq7"));
var _ut7q8es = _interopRequireDefault(require("./../../.stylex/ut7q8es"));
var _uugkpzq = _interopRequireDefault(require("./../../.stylex/uugkpzq"));
var _react = require("@universe-design/stylex/react");
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var React = _interopRequireWildcard(require("react"));
var _LeftOutlined = _interopRequireDefault(require("@universe-design/icons-react/LeftOutlined"));
var _RightOutlined = _interopRequireDefault(require("@universe-design/icons-react/RightOutlined"));
var _clsx = _interopRequireDefault(require("clsx"));
var _reactUtilities = require("@universe-design/react-utilities");
var _hooks = require("../../config-provider/hooks");
var _ImageViewerContext = require("../ImageViewerContext");
var _tooltip = _interopRequireDefault(require("../../tooltip"));
var _button = require("../../button");
var _space = require("../../space");
require("../style");
//.u22iagw{display:flex;}
//.u4d9j23{justify-content:center;}
//.uz5stix{white-space:nowrap;}
/* eslint-disable react/no-unknown-property */
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const usePaginationStylexStyles = (0, _react.makeStylexStyles)(_ref => {
  let {
    prefix
  } = _ref;
  return {
    info: {
      "props": {
        "mc9l5x": _ukchsq.default,
        "Brf1p80": _ut7q8es.default,
        "Huce71": _uugkpzq.default
      },
      "alias": `${prefix}__info`
    }
  };
});

/**
 * 图片查看器翻页工具 hook
 *
 * @public
 */
const useImageViewerToolbarPagination = () => {
  const {
    images,
    setCurrentIndex,
    currentIndex
  } = (0, _ImageViewerContext.useImageViewerContext)();
  const paginateForwardDisabled = currentIndex === images.length - 1;
  const paginateBackward = (0, _reactUtilities.useEventCallback)(() => {
    if (currentIndex > 0) {
      setCurrentIndex(currentIndex - 1);
    }
  });
  const paginateForward = (0, _reactUtilities.useEventCallback)(() => {
    if (currentIndex < images.length - 1) {
      setCurrentIndex(currentIndex + 1);
    }
  });
  React.useEffect(() => {
    const onKeydown = event => {
      if (event.key === _reactUtilities.KeyboardKey.ArrowLeft) {
        paginateBackward();
      } else if (event.key === _reactUtilities.KeyboardKey.ArrowRight) {
        paginateForward();
      }
    };
    document.addEventListener('keydown', onKeydown);
    return () => {
      document.removeEventListener('keydown', onKeydown);
    };
  }, [paginateBackward, paginateForward]);
  return {
    paginateBackwardDisabled: currentIndex === 0,
    paginateForwardDisabled,
    paginateBackward,
    paginateForward
  };
};
exports.useImageViewerToolbarPagination = useImageViewerToolbarPagination;
const Pagination = () => {
  const locale = (0, _hooks.useLocaleReceiver)('imageViewer');
  const {
    images,
    currentIndex,
    rootCssScope
  } = (0, _ImageViewerContext.useImageViewerContext)();
  const styles = usePaginationStylexStyles({
    prefix: rootCssScope('__pagination')
  });
  const {
    paginateBackward,
    paginateBackwardDisabled,
    paginateForward,
    paginateForwardDisabled
  } = useImageViewerToolbarPagination();
  return (0, _react.jsx)(_space.Space, {
    size: 12
  }, (0, _react.jsx)(_tooltip.default, {
    title: locale.prevArrow,
    overlayClassName: rootCssScope('__toolbar__tooltip'),
    wrapDisabledChildren: true
  }, (0, _react.jsx)(_button.IconButton, {
    size: 20,
    disabled: paginateBackwardDisabled,
    icon: (0, _react.jsx)(_LeftOutlined.default, null),
    className: (0, _clsx.default)(rootCssScope('__toolbar__button'), paginateBackwardDisabled && rootCssScope('__toolbar__button-disabled')),
    onClick: paginateBackward
  })), (0, _react.jsx)("div", {
    sx: [styles.info]
  }, `${currentIndex + 1} / ${images.length}`), (0, _react.jsx)(_tooltip.default, {
    title: locale.nextArrow,
    wrapDisabledChildren: true,
    overlayClassName: rootCssScope('__toolbar__tooltip')
  }, (0, _react.jsx)(_button.IconButton, {
    disabled: paginateForwardDisabled,
    size: 20,
    icon: (0, _react.jsx)(_RightOutlined.default, null),
    className: (0, _clsx.default)(rootCssScope('__toolbar__button'), paginateForwardDisabled && rootCssScope('__toolbar__button-disabled')),
    onClick: paginateForward
  })));
};
exports.Pagination = Pagination;
