// @universe-design/react/lib/image-viewer/components/Rotation.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useImageViewerToolbarRotation = exports.Rotation = void 0;
var React = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _RotateOutlined = _interopRequireDefault(require("@universe-design/icons-react/RotateOutlined"));
var _reactUtilities = require("@universe-design/react-utilities");
var _glMatrix = require("gl-matrix");
var _hooks = require("../../config-provider/hooks");
var _constants = require("../constants");
var _button = require("../../button");
var _tooltip = require("../../tooltip");
var _ImageViewerContext = require("../ImageViewerContext");
require("../style");
/**
 * 图片查看器旋转工具 hook
 *
 * @public
 */
const useImageViewerToolbarRotation = () => {
  const {
    getContainerRect,
    setMatrixByPoints
  } = (0, _ImageViewerContext.useImageViewerContext)();
  return {
    rotateCounterclockwise: (0, _reactUtilities.useEventCallback)(() => {
      const containerRect = getContainerRect();
      const origin = _glMatrix.vec2.fromValues((containerRect.x + containerRect.width) / 2, (containerRect.y + containerRect.height) / 2);
      const rotationMat = _glMatrix.mat2d.fromRotation(_glMatrix.mat2d.create(), _constants.ROTATION_RADIAN);
      setMatrixByPoints(absolutePoint => {
        const point = _glMatrix.vec2.subtract(_glMatrix.vec2.create(), absolutePoint, origin); // 基于 origin 的坐标
        let nextPoint = _glMatrix.vec2.create();
        nextPoint = _glMatrix.vec2.transformMat2d(_glMatrix.vec2.create(), point, rotationMat);
        nextPoint = _glMatrix.vec2.add(_glMatrix.vec2.create(), nextPoint, origin);
        return nextPoint;
      });
    })
  };
};
exports.useImageViewerToolbarRotation = useImageViewerToolbarRotation;
const Rotation = () => {
  const locale = (0, _hooks.useLocaleReceiver)('imageViewer');
  const {
    rootCssScope
  } = (0, _ImageViewerContext.useImageViewerContext)();
  const {
    rotateCounterclockwise
  } = useImageViewerToolbarRotation();
  return /*#__PURE__*/React.createElement(_tooltip.Tooltip, {
    title: locale.rotate,
    overlayClassName: rootCssScope('__toolbar__tooltip')
  }, /*#__PURE__*/React.createElement(_button.IconButton, {
    size: 20,
    icon: /*#__PURE__*/React.createElement(_RotateOutlined.default, null),
    className: (0, _clsx.default)(rootCssScope('__toolbar__button'), rootCssScope('__toolbar__rotation')),
    onClick: rotateCounterclockwise
  }));
};
exports.Rotation = Rotation;
