// @universe-design/react/lib/image-viewer/components/Scaling.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useImageViewerToolbarScaling = exports.Scaling = void 0;
var _uwf09ar = _interopRequireDefault(require("./../../.stylex/uwf09ar"));
var _ukchsq = _interopRequireDefault(require("./../../.stylex/ukchsq7"));
var _ut7q8es = _interopRequireDefault(require("./../../.stylex/ut7q8es"));
var _react = require("@universe-design/stylex/react");
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var React = _interopRequireWildcard(require("react"));
var _OriginalSizeOutlined = _interopRequireDefault(require("@universe-design/icons-react/OriginalSizeOutlined"));
var _ResizeOutlined = _interopRequireDefault(require("@universe-design/icons-react/ResizeOutlined"));
var _ZoomInOutlined = _interopRequireDefault(require("@universe-design/icons-react/ZoomInOutlined"));
var _ZoomOutOutlined = _interopRequireDefault(require("@universe-design/icons-react/ZoomOutOutlined"));
var _constants = require("../constants");
var _clsx = _interopRequireDefault(require("clsx"));
var _index = require("../../config-provider/hooks/index");
var _reactUtilities = require("@universe-design/react-utilities");
var _index2 = require("../../button/index");
var _index3 = require("../../tooltip/index");
var _index4 = require("../../hooks/index");
var _ImageViewerContext = require("../ImageViewerContext");
var _space = require("../../space");
require("../style/index");
//.ueqmc2u{width:40px;}
//.u22iagw{display:flex;}
//.u4d9j23{justify-content:center;}
/* eslint-disable @typescript-eslint/no-magic-numbers */
/* eslint-disable react/no-unknown-property */
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const useScalingStylexStyles = (0, _react.makeStylexStyles)(_ref => {
  let {
    prefix
  } = _ref;
  return {
    info: {
      "props": {
        "a9b677": _uwf09ar.default,
        "mc9l5x": _ukchsq.default,
        "Brf1p80": _ut7q8es.default
      },
      "alias": `${prefix}__info`
    }
  };
});

/**
 * 图片查看器缩放工具 hook
 *
 * @public
 */
const useImageViewerToolbarScaling = () => {
  // 从上下文获取图片查看器的缩放相关状态和方法
  const {
    scale,
    initTransform,
    scaleByOrigin,
    minScale,
    maxScale
  } = (0, _ImageViewerContext.useImageViewerContext)();

  // 适应屏幕状态，控制是否自动适配容器
  const [fitToScreen, setFitToScreen] = React.useState(true);

  // 切换适应屏幕状态的回调函数
  const toggleFitToScreen = (0, _index4.useEventCallback)(() => {
    setFitToScreen(!fitToScreen);
    initTransform({
      fitToContainer: !fitToScreen
    });
  });

  // 计算缩放按钮的禁用状态（保留两位小数避免浮点数精度问题）
  const scaleDownDisabled = Number(scale.toFixed(2)) <= minScale;
  const scaleUpDisabled = Number(scale.toFixed(2)) >= maxScale;

  // 放大操作处理函数
  const scaleUp = (0, _index4.useEventCallback)(() => {
    scaleByOrigin({
      getNextScale: currentScale => currentScale * (1 + _constants.SCALING_SPEED)
    });
  });

  // 缩小操作处理函数
  const scaleDown = (0, _index4.useEventCallback)(() => {
    scaleByOrigin({
      getNextScale: currentScale => currentScale * (1 - _constants.SCALING_SPEED)
    });
  });

  /**
   * Scaling
   * - 键盘操作（上下键）
   */
  React.useEffect(() => {
    /**
     * 上、下键缩放，基于容器中心缩放
     */
    const handleKeyDown = event => {
      if (event.key === _reactUtilities.KeyboardKey.ArrowUp) {
        scaleUp();
      } else if (event.key === _reactUtilities.KeyboardKey.ArrowDown) {
        scaleDown();
      }
    };
    document.addEventListener('keydown', handleKeyDown);
    return () => {
      document.removeEventListener('keydown', handleKeyDown);
    };
  }, [scaleDown, scaleUp]);
  return {
    fitToScreen,
    toggleFitToScreen,
    scaleDownDisabled,
    scaleUpDisabled,
    scaleUp,
    scaleDown
  };
};
exports.useImageViewerToolbarScaling = useImageViewerToolbarScaling;
const Scaling = () => {
  const locale = (0, _index.useLocaleReceiver)('imageViewer');
  const {
    scale,
    rootCssScope
  } = (0, _ImageViewerContext.useImageViewerContext)();
  const styles = useScalingStylexStyles({
    prefix: rootCssScope('__scaling')
  });
  const {
    fitToScreen,
    toggleFitToScreen,
    scaleDownDisabled,
    scaleUpDisabled,
    scaleUp,
    scaleDown
  } = useImageViewerToolbarScaling();
  return (0, _react.jsx)(_space.Space, {
    size: 12,
    className: rootCssScope('__scaling')
  }, (0, _react.jsx)(_index3.Tooltip, {
    title: locale.zoomOut,
    wrapDisabledChildren: true,
    overlayClassName: rootCssScope('__toolbar__tooltip')
  }, (0, _react.jsx)(_index2.IconButton, {
    size: 20,
    disabled: scaleDownDisabled,
    icon: (0, _react.jsx)(_ZoomOutOutlined.default, null),
    className: (0, _clsx.default)(rootCssScope('__toolbar__button'), scaleDownDisabled && rootCssScope('__toolbar__button-disabled')),
    onClick: scaleDown
  })), (0, _react.jsx)("div", {
    sx: [styles.info]
  }, `${(scale * _constants.AMPLIFY_NUMBER).toFixed(0)}%`), (0, _react.jsx)(_index3.Tooltip, {
    title: locale.zoomIn,
    wrapDisabledChildren: true,
    overlayClassName: rootCssScope('__toolbar__tooltip')
  }, (0, _react.jsx)(_index2.IconButton, {
    disabled: scaleUpDisabled,
    size: 20,
    icon: (0, _react.jsx)(_ZoomInOutlined.default, null),
    className: (0, _clsx.default)(rootCssScope('__toolbar__button'), scaleUpDisabled && rootCssScope('__toolbar__button-disabled')),
    onClick: scaleUp
  })), (0, _react.jsx)(_index3.Tooltip, {
    title: fitToScreen ? locale.originSize : locale.bestFit,
    overlayClassName: rootCssScope('__toolbar__tooltip')
  }, (0, _react.jsx)(_index2.IconButton, {
    size: 20,
    icon: fitToScreen ? (0, _react.jsx)(_OriginalSizeOutlined.default, null) : (0, _react.jsx)(_ResizeOutlined.default, null),
    className: rootCssScope('__toolbar__button'),
    onClick: toggleFitToScreen
  })));
};
exports.Scaling = Scaling;
