// @universe-design/styles/ts/ccm-chart-ig-color/index.light.atom.ts
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
export const ccm_chart_B100 = '#e1eaff';
export const ccm_chart_B200 = '#bacefd';
export const ccm_chart_B300 = '#82a7fc';
export const ccm_chart_B400 = '#4e83fd';
export const ccm_chart_B50 = '#f0f4ff';
export const ccm_chart_B500 = '#3370ff';
export const ccm_chart_B600 = '#245bdb';
export const ccm_chart_B700 = '#1c4cba';
export const ccm_chart_B800 = '#133c9a';
export const ccm_chart_B900 = '#0c296e';
export const ccm_chart_C100 = '#fdddef';
export const ccm_chart_C200 = '#f9aed9';
export const ccm_chart_C300 = '#f57ac0';
export const ccm_chart_C400 = '#f14ba9';
export const ccm_chart_C50 = '#fef1f8';
export const ccm_chart_C500 = '#f01d94';
export const ccm_chart_C600 = '#c71077';
export const ccm_chart_C700 = '#9e1361';
export const ccm_chart_C800 = '#7a0f4b';
export const ccm_chart_C900 = '#520a32';
export const ccm_chart_G100 = '#d9f5d6';
export const ccm_chart_G200 = '#b7edb1';
export const ccm_chart_G300 = '#8ee085';
export const ccm_chart_G400 = '#62d256';
export const ccm_chart_G50 = '#f0fbef';
export const ccm_chart_G500 = '#34c724';
export const ccm_chart_G600 = '#2ea121';
export const ccm_chart_G700 = '#237b19';
export const ccm_chart_G800 = '#186010';
export const ccm_chart_G900 = '#124b0c';
export const ccm_chart_I100 = '#e0e2fa';
export const ccm_chart_I200 = '#b4b9f3';
export const ccm_chart_I300 = '#7b83ea';
export const ccm_chart_I400 = '#616ae5';
export const ccm_chart_I50 = '#f2f3fd';
export const ccm_chart_I500 = '#4954e6';
export const ccm_chart_I600 = '#2933c7';
export const ccm_chart_I700 = '#161fa2';
export const ccm_chart_I800 = '#0a1285';
export const ccm_chart_I900 = '#0c1264';
export const ccm_chart_L100 = '#eef6c6';
export const ccm_chart_L200 = '#dfee96';
export const ccm_chart_L300 = '#d2e76a';
export const ccm_chart_L400 = '#c3dd40';
export const ccm_chart_L50 = '#f7fae5';
export const ccm_chart_L500 = '#b3d600';
export const ccm_chart_L600 = '#8fac02';
export const ccm_chart_L700 = '#667901';
export const ccm_chart_L800 = '#495700';
export const ccm_chart_L900 = '#354200';
export const ccm_chart_N00 = '#ffffff';
export const ccm_chart_N100 = '#f2f3f5';
export const ccm_chart_N1000 = '#000000';
export const ccm_chart_N200 = '#eff0f1';
export const ccm_chart_N300 = '#dee0e3';
export const ccm_chart_N400 = '#bbbfc4';
export const ccm_chart_N50 = '#f5f6f7';
export const ccm_chart_N500 = '#8f959e';
export const ccm_chart_N600 = '#646a73';
export const ccm_chart_N650 = '#51565d';
export const ccm_chart_N700 = '#373c43';
export const ccm_chart_N800 = '#2b2f36';
export const ccm_chart_N900 = '#1f2329';
export const ccm_chart_O100 = '#feead2';
export const ccm_chart_O200 = '#fed4a4';
export const ccm_chart_O300 = '#ffba6b';
export const ccm_chart_O400 = '#ffa53d';
export const ccm_chart_O50 = '#fff5eb';
export const ccm_chart_O500 = '#ff8800';
export const ccm_chart_O600 = '#de7802';
export const ccm_chart_O700 = '#b26206';
export const ccm_chart_O800 = '#8f4f04';
export const ccm_chart_O900 = '#6b3900';
export const ccm_chart_P100 = '#ece2fe';
export const ccm_chart_P200 = '#cdb2fa';
export const ccm_chart_P300 = '#ad82f7';
export const ccm_chart_P400 = '#935af6';
export const ccm_chart_P50 = '#f6f1fe';
export const ccm_chart_P500 = '#7f3bf5';
export const ccm_chart_P600 = '#6425d0';
export const ccm_chart_P700 = '#4e1ba7';
export const ccm_chart_P800 = '#380d82';
export const ccm_chart_P900 = '#270561';
export const ccm_chart_R100 = '#fde2e2';
export const ccm_chart_R200 = '#fbbfbc';
export const ccm_chart_R300 = '#f98e8b';
export const ccm_chart_R400 = '#f76964';
export const ccm_chart_R50 = '#fef1f1';
export const ccm_chart_R500 = '#f54a45';
export const ccm_chart_R600 = '#d83931';
export const ccm_chart_R700 = '#ac2f28';
export const ccm_chart_R800 = '#812520';
export const ccm_chart_R900 = '#621c18';
export const ccm_chart_T100 = '#d5f6f2';
export const ccm_chart_T200 = '#a9efe6';
export const ccm_chart_T300 = '#64e8d6';
export const ccm_chart_T400 = '#14e1c6';
export const ccm_chart_T50 = '#eafaf8';
export const ccm_chart_T500 = '#00d6b9';
export const ccm_chart_T600 = '#04b49c';
export const ccm_chart_T700 = '#078372';
export const ccm_chart_T800 = '#036356';
export const ccm_chart_T900 = '#024b41';
export const ccm_chart_V100 = '#f8def8';
export const ccm_chart_V200 = '#efb9ef';
export const ccm_chart_V300 = '#e58fe5';
export const ccm_chart_V400 = '#db66db';
export const ccm_chart_V50 = '#fceefc';
export const ccm_chart_V500 = '#d136d1';
export const ccm_chart_V600 = '#b220b2';
export const ccm_chart_V700 = '#8c218c';
export const ccm_chart_V800 = '#681268';
export const ccm_chart_V900 = '#460b46';
export const ccm_chart_W100 = '#d9f3fd';
export const ccm_chart_W200 = '#b1e8fc';
export const ccm_chart_W300 = '#7edafb';
export const ccm_chart_W400 = '#50cefb';
export const ccm_chart_W50 = '#ecf9fe';
export const ccm_chart_W500 = '#14c0ff';
export const ccm_chart_W600 = '#049fd7';
export const ccm_chart_W700 = '#037eaa';
export const ccm_chart_W800 = '#006185';
export const ccm_chart_W900 = '#004761';
export const ccm_chart_Y100 = '#faf1d1';
export const ccm_chart_Y200 = '#f8e6ab';
export const ccm_chart_Y300 = '#f7dc82';
export const ccm_chart_Y400 = '#fad355';
export const ccm_chart_Y50 = '#fdf9ed';
export const ccm_chart_Y500 = '#ffc60a';
export const ccm_chart_Y600 = '#dc9b04';
export const ccm_chart_Y700 = '#aa7803';
export const ccm_chart_Y800 = '#795101';
export const ccm_chart_Y900 = '#5c3a00';
export const ccm_chart_pareto_label_default = '#1f2329';
export const ccm_chart_pareto_line_default = '#e5ce00';
