// @universe-design/styles/ts/ccm-chart-ig-color/index.dark.atom.ts
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
export const ccm_chart_B100 = '#192a4c';
export const ccm_chart_B200 = '#203e78';
export const ccm_chart_B300 = '#2851a3';
export const ccm_chart_B400 = '#2e65d1';
export const ccm_chart_B50 = '#151f33';
export const ccm_chart_B500 = '#4c88ff';
export const ccm_chart_B600 = '#70a0ff';
export const ccm_chart_B700 = '#99bbff';
export const ccm_chart_B800 = '#b8cfff';
export const ccm_chart_B900 = '#cedfff';
export const ccm_chart_C100 = '#5c2542';
export const ccm_chart_C200 = '#91376a';
export const ccm_chart_C300 = '#ad407e';
export const ccm_chart_C400 = '#cc4792';
export const ccm_chart_C50 = '#421b31';
export const ccm_chart_C500 = '#e352a3';
export const ccm_chart_C600 = '#ed66b3';
export const ccm_chart_C700 = '#ff91cf';
export const ccm_chart_C800 = '#ffbde2';
export const ccm_chart_C900 = '#ffdbef';
export const ccm_chart_G100 = '#1f471b';
export const ccm_chart_G200 = '#296b22';
export const ccm_chart_G300 = '#339428';
export const ccm_chart_G400 = '#3bb02e';
export const ccm_chart_G50 = '#193016';
export const ccm_chart_G500 = '#54c248';
export const ccm_chart_G600 = '#6dd162';
export const ccm_chart_G700 = '#92de89';
export const ccm_chart_G800 = '#bdf0b8';
export const ccm_chart_G900 = '#d6f5d2';
export const ccm_chart_I100 = '#262a5e';
export const ccm_chart_I200 = '#353b8c';
export const ccm_chart_I300 = '#4149b5';
export const ccm_chart_I400 = '#4c55d4';
export const ccm_chart_I50 = '#1d1f47';
export const ccm_chart_I500 = '#5f69ed';
export const ccm_chart_I600 = '#737beb';
export const ccm_chart_I700 = '#9ea5ff';
export const ccm_chart_I800 = '#bbc0f6';
export const ccm_chart_I900 = '#d2d5fd';
export const ccm_chart_L100 = '#495706';
export const ccm_chart_L200 = '#6d8205';
export const ccm_chart_L300 = '#91ad02';
export const ccm_chart_L400 = '#abc912';
export const ccm_chart_L50 = '#2e3605';
export const ccm_chart_L500 = '#b9d627';
export const ccm_chart_L600 = '#c7e04d';
export const ccm_chart_L700 = '#d6ea75';
export const ccm_chart_L800 = '#e7f5a6';
export const ccm_chart_L900 = '#f6ffcc';
export const ccm_chart_N00 = '#0a0a0a';
export const ccm_chart_N100 = '#292929';
export const ccm_chart_N1000 = '#ffffff';
export const ccm_chart_N200 = '#373737';
export const ccm_chart_N300 = '#434343';
export const ccm_chart_N400 = '#5f5f5f';
export const ccm_chart_N50 = '#1a1a1a';
export const ccm_chart_N500 = '#757575';
export const ccm_chart_N600 = '#a6a6a6';
export const ccm_chart_N650 = '#cfcfcf';
export const ccm_chart_N700 = '#e0e0e0';
export const ccm_chart_N800 = '#e8e8e8';
export const ccm_chart_N900 = '#ebebeb';
export const ccm_chart_O100 = '#57330a';
export const ccm_chart_O200 = '#845117';
export const ccm_chart_O300 = '#bd7017';
export const ccm_chart_O400 = '#de8218';
export const ccm_chart_O50 = '#33210b';
export const ccm_chart_O500 = '#f2962c';
export const ccm_chart_O600 = '#f5a54a';
export const ccm_chart_O700 = '#fabc75';
export const ccm_chart_O800 = '#ffd8ac';
export const ccm_chart_O900 = '#ffeace';
export const ccm_chart_P100 = '#361d61';
export const ccm_chart_P200 = '#4d2691';
export const ccm_chart_P300 = '#6533bd';
export const ccm_chart_P400 = '#7c4ad4';
export const ccm_chart_P50 = '#281745';
export const ccm_chart_P500 = '#9762f5';
export const ccm_chart_P600 = '#a472fc';
export const ccm_chart_P700 = '#b990ff';
export const ccm_chart_P800 = '#d3bafd';
export const ccm_chart_P900 = '#e3d4ff';
export const ccm_chart_R100 = '#4a1d1b';
export const ccm_chart_R200 = '#6b2827';
export const ccm_chart_R300 = '#9e3836';
export const ccm_chart_R400 = '#cc4743';
export const ccm_chart_R50 = '#361918';
export const ccm_chart_R500 = '#f05b56';
export const ccm_chart_R600 = '#fa7873';
export const ccm_chart_R700 = '#ff9c99';
export const ccm_chart_R800 = '#fec6c3';
export const ccm_chart_R900 = '#ffe0e0';
export const ccm_chart_T100 = '#153d38';
export const ccm_chart_T200 = '#23665d';
export const ccm_chart_T300 = '#269687';
export const ccm_chart_T400 = '#37b8a6';
export const ccm_chart_T50 = '#162e2b';
export const ccm_chart_T500 = '#41ccb9';
export const ccm_chart_T600 = '#5cdbca';
export const ccm_chart_T700 = '#7febdc';
export const ccm_chart_T800 = '#b7f7ef';
export const ccm_chart_T900 = '#daf9f5';
export const ccm_chart_V100 = '#4f1c4f';
export const ccm_chart_V200 = '#7a287a';
export const ccm_chart_V300 = '#943b94';
export const ccm_chart_V400 = '#b54ab5';
export const ccm_chart_V50 = '#3b153b';
export const ccm_chart_V500 = '#d457d4';
export const ccm_chart_V600 = '#de73de';
export const ccm_chart_V700 = '#e898e8';
export const ccm_chart_V800 = '#f2bff2';
export const ccm_chart_V900 = '#fbdafb';
export const ccm_chart_W100 = '#173742';
export const ccm_chart_W200 = '#225b70';
export const ccm_chart_W300 = '#2782a3';
export const ccm_chart_W400 = '#2ea5d1';
export const ccm_chart_W50 = '#142930';
export const ccm_chart_W500 = '#42bdeb';
export const ccm_chart_W600 = '#5fd4fe';
export const ccm_chart_W700 = '#89dffe';
export const ccm_chart_W800 = '#c2eeff';
export const ccm_chart_W900 = '#def6ff';
export const ccm_chart_Y100 = '#574711';
export const ccm_chart_Y200 = '#8f7211';
export const ccm_chart_Y300 = '#c99d08';
export const ccm_chart_Y400 = '#f0bb0a';
export const ccm_chart_Y50 = '#382e0c';
export const ccm_chart_Y500 = '#fac823';
export const ccm_chart_Y600 = '#fcd456';
export const ccm_chart_Y700 = '#ffde75';
export const ccm_chart_Y800 = '#ffe694';
export const ccm_chart_Y900 = '#fcefc2';
export const ccm_chart_pareto_label_default = '#ebebeb';
export const ccm_chart_pareto_line_default = '#faed57';
