// @universe-design/react/lib/anchor/AnchorLink.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AnchorLink = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _stylex2 = require("../stylex");
var _util = require("../util");
var _hooks = require("../config-provider/hooks");
var _hooks2 = require("../hooks");
var _AnchorContext = require("./AnchorContext");
require("./style");
/**
 * Anchor.Link 组件的属性
 * @public
 */

const AnchorLink = exports.AnchorLink = /*#__PURE__*/_react.default.forwardRef((props, ref) => {
  const {
    href,
    title,
    target,
    children,
    className,
    onClick,
    ...restProps
  } = props;
  const cssScope = (0, _hooks.useCssScope)('anchor-link');
  const {
    activeLink,
    registerLink,
    unregisterLink,
    handleLinkClick
  } = (0, _AnchorContext.useAnchorContext)();
  const linkRef = _react.default.useRef(null);
  const cls = (0, _clsx.default)(cssScope(''), className, {
    [cssScope('--active')]: activeLink === href
  });
  const titleCls = (0, _clsx.default)(cssScope('__title'), {
    [cssScope('__title--active')]: activeLink === href
  }, (0, _stylex.default)(_stylex2.cssTypographyBody0, activeLink === href && _stylex2.cssTypographyHeadline));
  (0, _hooks2.useIsomorphicLayoutEffect)(() => {
    registerLink(href, {
      ref: linkRef
    });
    return () => unregisterLink(href);
  }, [href, registerLink, unregisterLink]);
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
    className: cls,
    ref: ref
  }, restProps), /*#__PURE__*/_react.default.createElement("a", {
    ref: linkRef,
    className: titleCls,
    href: href,
    title: typeof title === 'string' ? title : '',
    target: target,
    onClick: e => {
      (0, _util.isFunction)(onClick) && onClick(e, {
        title,
        href
      });
      handleLinkClick(e, {
        title,
        href
      });
    }
  }, title), children);
});
if (process.env.NODE_ENV !== "production") {
  AnchorLink.displayName = 'AnchorLink';
}
