// @universe-design/react/lib/util/datastructure/LinkedList.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LinkedListItem = exports.LinkedList = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
/* eslint-disable max-depth */
class LinkedListItem {
  constructor(val) {
    (0, _defineProperty2.default)(this, "value", void 0);
    (0, _defineProperty2.default)(this, "next", void 0);
    (0, _defineProperty2.default)(this, "prev", void 0);
    this.value = val;
    this.next = null;
    this.prev = null;
  }
}
exports.LinkedListItem = LinkedListItem;
class LinkedList {
  constructor() {
    let values = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
    (0, _defineProperty2.default)(this, "_head", void 0);
    (0, _defineProperty2.default)(this, "_tail", void 0);
    (0, _defineProperty2.default)(this, "_length", void 0);
    this._head = null;
    this._tail = null;
    this._length = 0;
    if (values.length > 0) {
      values.forEach(value => {
        this.append(value);
      });
    }
  }
  get head() {
    return this._head ? this._head.value : null;
  }
  get tail() {
    return this._tail ? this._tail.value : null;
  }
  get length() {
    return this._length;
  }
  *iterator() {
    let currentItem = this._head;
    while (currentItem) {
      yield currentItem.value;
      currentItem = currentItem.next;
    }
  }
  [Symbol.iterator]() {
    return this.iterator();
  }

  // Adds the element at the end of the linked list
  append(val) {
    let checkDuplicates = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
    if (checkDuplicates && this.isDuplicate(val)) {
      return false;
    }
    const newItem = new LinkedListItem(val);
    if (!this._tail) {
      this._head = newItem;
      this._tail = newItem;
    } else {
      this._tail.next = newItem;
      newItem.prev = this._tail;
      this._tail = newItem;
    }
    this._length++;
    return true;
  }

  // Add the element at the beginning of the linked list
  prepend(val) {
    let checkDuplicates = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
    if (checkDuplicates && this.isDuplicate(val)) {
      return false;
    }
    const newItem = new LinkedListItem(val);
    if (!this._head) {
      this._head = newItem;
      this._tail = newItem;
    } else {
      newItem.next = this._head;
      this._head.prev = newItem;
      this._head = newItem;
    }
    this._length++;
    return true;
  }
  removeHead() {
    var _this$_head;
    const currentItem = this._head;

    // empty list
    if (!currentItem) {
      return null;
    }

    // single item list
    if (!((_this$_head = this._head) !== null && _this$_head !== void 0 && _this$_head.next)) {
      this._head = null;
      this._tail = null;

      // full list
    } else {
      this._head.next.prev = null;
      this._head = this._head.next;
      currentItem.next = null;
      currentItem.prev = null;
    }
    this._length--;
    return currentItem.value;
  }
  removeTail() {
    var _this$_tail;
    const currentItem = this._tail;

    // empty list
    if (!currentItem) {
      return null;
    }

    // single item list
    if (!((_this$_tail = this._tail) !== null && _this$_tail !== void 0 && _this$_tail.prev)) {
      this._head = null;
      this._tail = null;

      // full list
    } else {
      this._tail.prev.next = null;
      this._tail = this._tail.prev;
      currentItem.next = null;
      currentItem.prev = null;
    }
    this._length--;
    return currentItem.value;
  }
  first(num) {
    const iter = this.iterator();
    const result = [];
    const n = Math.min(num, this.length);
    for (let i = 0; i < n; i++) {
      const val = iter.next();
      result.push(val.value);
    }
    return result;
  }
  toArray() {
    return [...this];
  }
  isDuplicate(val) {
    const set = new Set(this.toArray());
    return set.has(val);
  }
}
exports.LinkedList = LinkedList;
