// @universe-design/react/src/tree/hooks/checkStrategy.ts
import { type TKeyValueMap } from '../../util';
import { CheckedType, ShowCheckedStrategy } from '../interface';
import type { DoublyTreeNode } from '../util';
import { isNonLeaf2 } from '../util';
import { getCheckedCache } from './getCheckedCache';

export type CheckedTypeCache = Map<string, CheckedType>;

// 尽可能用指针来减少 key 太长带来的内存问题
export interface CheckStrategy {
  /**
   * 得到 onChange 输出的结果
   */
  cache2CheckedSet: (doublyTreeNodeMap: TKeyValueMap<DoublyTreeNode>, cache: CheckedTypeCache) => TKeyValueMap<boolean>;
  /**
   * 从 checkedSet 得到 chache 结果
   */
  checkedSet2Cache: (
    doublyTreeNodeMap: TKeyValueMap<DoublyTreeNode>,
    checkedSet: TKeyValueMap<boolean>,
  ) => CheckedTypeCache;
}

/**
 * show all
 */
export const checkStrategyShowAll: CheckStrategy = {
  cache2CheckedSet(doublyTreeNodeMap, cache) {
    const checkedSet: TKeyValueMap<boolean> = {};
    for (const [k, v] of cache.entries()) {
      checkedSet[k] = v === CheckedType.Checked;
    }
    return checkedSet;
  },
  checkedSet2Cache(doublyTreeNodeMap, checkedSet) {
    const checkedKeys = Object.entries(checkedSet)
      .filter(([, v]) => v === true)
      .map(([x]) => x);
    return getCheckedCache({
      doublyTreeNodeMap,
      checkedKeys,
      checkStrictly: false,
    });
  },
};

/**
 * show parent
 * 如果所有的 children 都选中则只显示 parent
 */
export const checkStrategyShowParent: CheckStrategy = {
  ...checkStrategyShowAll,
  cache2CheckedSet(doublyTreeNodeMap, cache) {
    const checkedSet: TKeyValueMap<boolean> = {};
    // 补充不存在的 value
    for (const [key, checkedType] of cache) {
      if (checkedType === CheckedType.Checked) {
        checkedSet[key] = true;
        const node = doublyTreeNodeMap[key];
        const allSiblingsChecked = node?.parent?.children?.every(
          _node => cache.get(_node.__node__.value) === CheckedType.Checked,
          //  || isCheckedDisabled(_node.__node__),
        );
        checkedSet[key] = !allSiblingsChecked;
      } else {
        checkedSet[key] = false;
      }
    }

    return checkedSet;
  },
};

/**
 * show children
 * 如果所有的 children 都只显示 parent
 */
export const checkStrategyShowChild: CheckStrategy = {
  ...checkStrategyShowAll,
  cache2CheckedSet(doublyTreeNodeMap, cache) {
    const checkedSet: TKeyValueMap<boolean> = {};

    for (const [key, checkedType] of cache.entries()) {
      const node = doublyTreeNodeMap[key];
      // 可能 node 不存在，此时仍然将选择态保留
      const isChild = !node || !isNonLeaf2(node);
      checkedSet[key] = checkedType === CheckedType.Checked && isChild;
    }
    return checkedSet;
  },
};

export const strategyMap = {
  [ShowCheckedStrategy.SHOW_ALL]: checkStrategyShowAll,
  [ShowCheckedStrategy.SHOW_CHILD]: checkStrategyShowChild,
  [ShowCheckedStrategy.SHOW_PARENT]: checkStrategyShowParent,
};
