// @universe-design/react/src/tree/hooks/getCheckedCache.ts
import type { TKeyValueMap } from '../../util';
import type { CheckStrictlyKeys } from '../interface';
import { CheckedType } from '../interface';
import { type DoublyTreeNode } from '../util';
// import { isCheckedDisabled, type DoublyTreeNode } from '../util';
export type CheckedTypeCache = Map<string, CheckedType>;

type NodesByDepth = Map<number, DoublyTreeNode[]>;

function groupNodePyDepth(doublyTreeNodeMap: TKeyValueMap<DoublyTreeNode>) {
  const nodesByDepth: NodesByDepth = new Map();
  Object.values(doublyTreeNodeMap).forEach(node => {
    let nodeList = nodesByDepth.get(node.depth);
    if (!nodeList) {
      nodeList = [node];
      nodesByDepth.set(node.depth, nodeList);
    } else {
      nodeList.push(node);
    }
  });
  return nodesByDepth;
}

export function getCheckedCache(
  param: {
    doublyTreeNodeMap: TKeyValueMap<DoublyTreeNode>;
    keyToRemove?: string;
    keyToAdd?: string;
    checkStrictly: boolean;
  } & (
    | {
        checkStrictly: true;
        checkedKeys: CheckStrictlyKeys;
      }
    | {
        checkStrictly: false;
        checkedKeys: string[];
      }
  ),
): CheckedTypeCache {
  const { keyToRemove, keyToAdd, checkedKeys: checkedKeysProp, doublyTreeNodeMap, checkStrictly } = param;
  const checkedCache: CheckedTypeCache = new Map();

  /**
   *  checkStrictly 的情况
   */
  if (checkStrictly) {
    const { checked, halfChecked } = checkedKeysProp as CheckStrictlyKeys;
    checked.forEach(k => {
      if (k !== keyToRemove) {
        checkedCache.set(k, CheckedType.Checked);
      }
    });

    halfChecked.forEach(k => checkedCache.set(k, CheckedType.Indeterminate));

    if (keyToAdd !== undefined) {
      checkedCache.set(keyToAdd, CheckedType.Checked);
    }

    Object.keys(doublyTreeNodeMap).forEach(key => {
      if (checkedCache.get(key) === undefined) {
        checkedCache.set(key, CheckedType.Unchecked);
      }
    });
    return checkedCache;
  }

  /**
   * 非 checkStrictly 的情况
   */

  let keysToRemove = keyToRemove === undefined ? [] : [keyToRemove];

  const checkedKeys = checkedKeysProp as string[];

  checkedKeys.forEach(key => {
    checkedCache.set(key, CheckedType.Checked);
  });
  if (keyToAdd !== undefined) {
    checkedCache.set(keyToAdd, CheckedType.Checked);
  }

  const isNodeChecked = (nodeKey: string) => checkedCache.get(nodeKey) === CheckedType.Checked;
  const isNodeUnChecked = (nodeKey: string) =>
    !checkedCache.has(nodeKey) || checkedCache.get(nodeKey) === CheckedType.Unchecked;
  const isNodeHalfChecked = (nodeKey: string) => checkedCache.get(nodeKey) === CheckedType.Indeterminate;

  const nodesByDepth = groupNodePyDepth(doublyTreeNodeMap);

  const maxDepth = nodesByDepth.size;
  /* from root to leaf
   * checked 有传导性
   * 从上向下，如果父节点选中，则将子节点置为选中态
   */
  for (let depth = 1; depth <= maxDepth; depth++) {
    const nodes = nodesByDepth.get(depth);
    if (!nodes) {
      continue;
    }
    // eslint-disable-next-line @typescript-eslint/no-loop-func
    nodes.forEach(node => {
      const parentKey = node.parent?.key;
      const nodeKey = node.key;

      // if (isCheckedDisabled(node.__node__)) {
      //   return;
      // }

      if (keysToRemove.includes(nodeKey)) {
        checkedCache.set(nodeKey, CheckedType.Unchecked);

        /**
         * 删除需要递归删除子节点
         */
        if (node.children) {
          keysToRemove = keysToRemove.concat(node.children.map(n => n.key));
        }
        return;
      }

      if (parentKey !== undefined && checkedCache.has(parentKey)) {
        checkedCache.set(nodeKey, CheckedType.Checked);
      }
    });
  }

  /* from leaf to node
   * checked 有传导性
   * 如果子节点都选中，则父节点选中
   * 如果子节点都未选中，则父节点未选中
   * 否则父节点处于半选状态
   */
  for (let depth = maxDepth; depth >= 1; depth--) {
    const nodes = nodesByDepth.get(depth);
    if (!nodes) {
      continue;
    }
    nodes.forEach(node => {
      /**
       *  从子节点向上遍历时，如果子节点未选中，已经可以认定其最终状态是未选中了
       **/
      const nodeKey = node.key;
      if (!checkedCache.has(nodeKey)) {
        checkedCache.set(nodeKey, CheckedType.Unchecked);
      }

      // if (isCheckedDisabled(node.__node__)) {
      //   return;
      // }

      const parentKey = node.__parent__?.value;
      if (parentKey === undefined) {
        return;
      }

      let isAllSiblingsChecked = true;
      let isAllSiblingsUnChecked = true;
      let isAllSiblingsDisabled = true;

      const parentNode = doublyTreeNodeMap[parentKey];
      // if (isCheckedDisabled(parentNode.__node__)) {
      //   return;
      // }
      const siblings = parentNode.children || [];
      siblings.forEach(sibling => {
        // if (isCheckedDisabled(sibling.__node__)) {
        //   return;
        // }

        isAllSiblingsDisabled = false;
        if (isNodeHalfChecked(sibling.key)) {
          isAllSiblingsUnChecked = false;
          isAllSiblingsChecked = false;
        }
        if (isNodeChecked(sibling.key)) {
          isAllSiblingsUnChecked = false;
        } else if (isNodeUnChecked(sibling.key)) {
          isAllSiblingsChecked = false;
        }
      });

      if (isAllSiblingsDisabled) {
        return;
      }

      if (isAllSiblingsUnChecked) {
        checkedCache.set(parentKey, CheckedType.Unchecked);
      } else if (isAllSiblingsChecked) {
        checkedCache.set(parentKey, CheckedType.Checked);
      } else {
        checkedCache.set(parentKey, CheckedType.Indeterminate);
      }
    });
  }
  return checkedCache;
}
