// @universe-design/react/src/tree/hooks/useTree.tsx
import * as React from 'react';
import { useControlledState } from '@universe-design/react-utilities';
import { treeChildren2TreeNodes, useTreeCssScope } from '../common';
import type { TreeProps } from '../interface';
import { ShowCheckedStrategy } from '../interface';
import type { TreeContextType } from './TreeContext';
import { TreeContextProvider } from './TreeContext';
import { useTreeSelect } from './useTreeSelect';
import { useTreeCheck } from './useTreeCheck';
import { useTreeExpand } from './useTreeExpand';
import { strategyMap } from './checkStrategy';
import { genDoublyTreeNodeMap } from '../util';
import type { ListRef } from '@universe-design/rc-virtual-list';
import { useLocale } from '../../config-provider';
import type { ListScrollConfig } from '../../util';

export type TreeListItemReference = number | { type: 'node'; key: string };

/**
 * @docgen
 */
export interface TreeRef {
  /**
   * 切换展开状态
   */
  toggleExpanded: ReturnType<typeof useTreeExpand>['toggleExpanded'];
  /**
   * 将指定选项滚动到可视区域内
   */
  scrollIntoView: (reference: TreeListItemReference, scrollConfig?: ListScrollConfig) => void;
}

const defaultFilterTreeNode = () => true;

export const useTree = (props: TreeProps, ref: React.Ref<TreeRef>): TreeContextType => {
  const locale = useLocale();
  const childrenToTreeData = React.useMemo(() => treeChildren2TreeNodes(props.children), [props.children]);
  const {
    treeData = childrenToTreeData,
    defaultExpandedKeys = [],
    expandedKeys,
    onExpand,
    defaultSelectedKeys = [],
    selectedKeys,
    onSelect,
    filterTreeNode = defaultFilterTreeNode,
    showChecker = false,
    checkable = false,
    selectable = true,
    defaultExpandAll = false,
    multiple,
    showCheckedStrategy = ShowCheckedStrategy.SHOW_ALL,
    switcherIcon,
    labelRenderer,
    loadData,
    onLoad,
    loadedKeys: loadedKeysProp,
    searchValue,
    notFoundContent = locale.global.notFoundContent,
    // virtual scrolling
    virtual = true,
    height,
    nodeRenderer,
  } = props;
  const cssScope = useTreeCssScope();
  const [loadedKeys, setLoadedKeys] = useControlledState<string[]>([], loadedKeysProp);

  // ======== 创建一个带索引的 tree node 双向链表 ========
  const { doublyTreeNodeList, doublyTreeNodeMap } = React.useMemo(() => genDoublyTreeNodeMap(treeData), [treeData]);

  const { finalTreeNodeList, filteredTreeNodes, expandedMap, toggleExpanded, finalTreeNodeListRef } = useTreeExpand({
    doublyTreeNodeMap,
    doublyTreeNodeList,
    defaultExpandAll,
    defaultExpandedKeys,
    expandedKeys,
    onExpand,
    filterTreeNode,
    searchValue,
  });

  const treeCheckCtx = useTreeCheck({
    doublyTreeNodeMap,
    checkStrategy: strategyMap[showCheckedStrategy],
    ...(props.checkStrictly === true
      ? {
          ...props,
          checkStrictly: true,
        }
      : { ...props, checkStrictly: false }),
  });

  const treeSelectCtx = useTreeSelect({
    doublyTreeNodeMap,
    defaultSelectedKeys,
    selectedKeys,
    onSelect,
    multiple,
  });

  // =========================== List ===========================
  const listRef = React.useRef<ListRef>(null);

  const scrollIntoView = React.useCallback(
    (reference: TreeListItemReference, scrollConfig?: ListScrollConfig) => {
      if (!listRef.current) {
        return;
      }
      if (typeof reference === 'number') {
        listRef.current.scrollTo({ index: reference, ...scrollConfig });
        return;
      }
      const { key } = reference;
      const parentKey = doublyTreeNodeMap[key]?.__parent__?.value;
      if (parentKey && !expandedMap[parentKey]) {
        toggleExpanded(parentKey);
      }
      /**
       * 这里需要一个异步处理：
       * 等待节点展开、下一次 render 之后，才能拿到更新后的 finalTreeNodeList
       * */
      setTimeout(() => {
        const index = finalTreeNodeListRef.current.findIndex(n => n.key === key);
        if (index !== -1) {
          listRef.current?.scrollTo({ index, ...scrollConfig });
        }
      });
    },
    [expandedMap, listRef, finalTreeNodeListRef, toggleExpanded, doublyTreeNodeMap],
  );

  React.useImperativeHandle(
    ref,
    () => ({
      toggleExpanded,
      scrollIntoView,
    }),
    [scrollIntoView, toggleExpanded],
  );

  return React.useMemo(
    () => ({
      loadData,
      onLoad,
      loadedKeys,
      selectable,
      checkable,
      setLoadedKeys,
      rootCssScope: cssScope,
      nodeRenderer,
      showChecker,
      expandedMap,
      toggleExpanded,
      switcherIcon,
      finalTreeNodeList,
      labelRenderer,
      filteredTreeNodes,
      ...treeCheckCtx,
      ...treeSelectCtx,
      height,
      virtual,
      listRef,
      searchValue,
      notFoundContent,
    }),
    [
      checkable,
      cssScope,
      expandedMap,
      filteredTreeNodes,
      finalTreeNodeList,
      labelRenderer,
      loadData,
      loadedKeys,
      nodeRenderer,
      onLoad,
      selectable,
      setLoadedKeys,
      showChecker,
      switcherIcon,
      toggleExpanded,
      treeCheckCtx,
      treeSelectCtx,
      height,
      virtual,
      searchValue,
      notFoundContent,
    ],
  );
};

export const TreeProvider = React.forwardRef<TreeRef, TreeProps>(({ children, ...props }, ref) => {
  const treeCtxValue = useTree(props, ref);
  return <TreeContextProvider value={treeCtxValue}>{children}</TreeContextProvider>;
});
