// @universe-design/react/src/tree/hooks/useTreeCheck.ts
import { useControlledState } from '@universe-design/react-utilities';
import React from 'react';
import type { Required } from 'utility-types';
import type { TKeyValueMap } from '../../util';
import { genArray, isFunction } from '../../util';
import type { CheckStrictlyKeys, TreeCheckProps, TreeCheckStricklyProps } from '../interface';
import { CheckedType } from '../interface';
import type { DoublyTreeNode } from '../util';
import { array2Set, isCheckedDisabled } from '../util';
import type { ITreeCheckedContext } from './TreeCheckedContext';
import { useTreeCheckedContext } from './TreeCheckedContext';
import type { CheckStrategy } from './checkStrategy';
import { getCheckedCache } from './getCheckedCache';

export function useTreeCheck(
  props: {
    doublyTreeNodeMap: TKeyValueMap<DoublyTreeNode>;
    checkStrategy: CheckStrategy;
  } & (Required<TreeCheckProps, 'checkStrictly'> | TreeCheckStricklyProps),
  { skipped: skippedOpt = false }: { skipped?: boolean } = {},
): ITreeCheckedContext {
  const { doublyTreeNodeMap, checkStrategy, checkStrictly, defaultCheckedKeys, checkedKeys: checkedKeysProp } = props;

  // 这段逻辑保证单独使用 useTreeCheck 也一定没有问题
  const treeValueContext = useTreeCheckedContext();
  const skipped = Boolean(treeValueContext) || skippedOpt;

  // eslint-disable-next-line prefer-const
  let [checkedKeys, setCheckedKeys] = useControlledState(
    defaultCheckedKeys ?? (checkStrictly ? { halfChecked: [], checked: [] } : []),
    checkedKeysProp,
    {
      isControlledOverride: 'checkedKeys' in props,
    },
  );

  let checkedCache = React.useMemo(() => {
    if (skipped) {
      return new Map();
    }

    if (checkStrictly) {
      const cache = new Map<string, CheckedType>();
      const { checked, halfChecked } = checkedKeys as CheckStrictlyKeys;
      checked.forEach(k => cache.set(k, CheckedType.Checked));
      halfChecked.forEach(k => cache.set(k, CheckedType.Indeterminate));
      Object.keys(doublyTreeNodeMap).forEach(key => {
        if (cache.get(key) === undefined) {
          cache.set(key, CheckedType.Unchecked);
        }
      });
      return cache;
    } else {
      const keys = genArray<string>(checkedKeys as string[]);
      return checkStrategy.checkedSet2Cache(doublyTreeNodeMap, array2Set(keys));
    }
  }, [checkStrategy, checkStrictly, checkedKeys, doublyTreeNodeMap, skipped]);

  const checkedSet = React.useMemo(() => {
    if (skipped) {
      return {};
    }
    if (checkStrictly) {
      const r: TKeyValueMap<boolean> = {};
      for (const [k, v] of checkedCache.entries()) {
        if (v === CheckedType.Checked) {
          r[k] = true;
        }
      }
      return r;
    }

    return checkStrategy.cache2CheckedSet(doublyTreeNodeMap, checkedCache);
  }, [skipped, checkStrictly, checkStrategy, doublyTreeNodeMap, checkedCache]);

  if (treeValueContext) {
    ({ checkedCache, setCheckedKeys } = treeValueContext);
  }

  /**
   * 切换 checked 状态
   */
  const toggleChecked = React.useCallback(
    (key: string) => {
      const node = doublyTreeNodeMap[key];
      if (!node || isCheckedDisabled(node.__node__)) {
        return;
      }

      const checkedType = checkedCache.get(node.__node__.value);

      let keyToRemove: string | undefined;
      let keyToAdd: string | undefined;

      if (checkedType === CheckedType.Checked) {
        keyToRemove = node.key;
      } else {
        keyToAdd = node.key;
      }

      if (props.checkStrictly) {
        const initialValue: CheckStrictlyKeys = {
          checked: [],
          halfChecked: [],
        };
        const checkedCache = getCheckedCache({
          doublyTreeNodeMap,
          checkedKeys: checkedKeys as CheckStrictlyKeys,
          keyToAdd,
          keyToRemove,
          checkStrictly: true,
        });

        const newCheckedKeys = initialValue;
        for (const [k, v] of checkedCache.entries()) {
          if (v === CheckedType.Checked) {
            newCheckedKeys.checked.push(k);
          } else if (v === CheckedType.Indeterminate) {
            newCheckedKeys.halfChecked.push(k);
          }
        }
        setCheckedKeys(
          newCheckedKeys,
          () =>
            isFunction(props.onCheck) &&
            props.onCheck(newCheckedKeys as CheckStrictlyKeys, {
              checked: checkedType !== CheckedType.Checked,
              key,
            }),
        );
      } else {
        const checkedCache = getCheckedCache({
          doublyTreeNodeMap,
          checkedKeys: checkedKeys as string[],
          keyToAdd,
          keyToRemove,
          checkStrictly: false,
        });

        const checkedMap = checkStrategy.cache2CheckedSet(doublyTreeNodeMap, checkedCache);
        const newKeys = Object.keys(checkedMap).filter(k => checkedMap[k]);
        setCheckedKeys(
          newKeys,
          () =>
            isFunction(props.onCheck) &&
            props.onCheck(newKeys, {
              checked: checkedType !== CheckedType.Checked,
              key,
            }),
        );
      }
    },
    [doublyTreeNodeMap, checkedCache, props, checkedKeys, setCheckedKeys, checkStrategy],
  );

  return React.useMemo(
    () => ({ checkedCache, checkedSet, setCheckedKeys, toggleChecked }),
    [checkedCache, checkedSet, setCheckedKeys, toggleChecked],
  );
}
