// @universe-design/react/src/space/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import CAlign from '../demo/align';
import CBase from '../demo/base';
import CCustomize from '../demo/customize';
import CSize from '../demo/size';
import CSplit from '../demo/split';
import CVertical from '../demo/vertical';
import CWrap from '../demo/wrap';
if (!(CAlign as Story).storyName) {
  (CAlign as Story).storyName = '4. 对齐';
  (CAlign as Story).parameters = {
    docs: {
      description: {
        story:
          '设置对齐模式。\n\n\n<style>\n.space-align-container {\n  display: flex;\n  align-item: flex-start;\n  flex-wrap: wrap;\n}\n.space-align-block {\n  margin: 8px 8px;\n  border: 1px solid #40a9ff;\n  padding: 4px;\n  flex: none;\n  border-color: #3370ff;\n}\n.space-align-block .mock-block {\n  display: inline-block;\n  padding: 32px 8px 16px;\n  background: rgba(150, 150, 150, 0.2);\n}\n</style>\n',
      },
    },
  };
}
if (!(CBase as Story).storyName) {
  (CBase as Story).storyName = '0. 基础用法';
  (CBase as Story).parameters = {
    docs: {
      description: {
        story: '间距组件的基础用法。\n\n',
      },
    },
  };
}
if (!(CCustomize as Story).storyName) {
  (CCustomize as Story).storyName = '3. 自定义尺寸';
  (CCustomize as Story).parameters = {
    docs: {
      description: {
        story: '自定义间距大小。\n\n',
      },
    },
  };
}
if (!(CSize as Story).storyName) {
  (CSize as Story).storyName = '2. 间距大小';
  (CSize as Story).parameters = {
    docs: {
      description: {
        story: '设置`size`可以改变间距大小。共有 3 种间距，分别为`large`、`medium`、`small`，默认为`medium`。\n\n',
      },
    },
  };
}
if (!(CSplit as Story).storyName) {
  (CSplit as Story).storyName = '99. 分隔符';
  (CSplit as Story).parameters = {
    docs: {
      description: {
        story: '相邻组件分隔符。\n\n',
      },
    },
  };
}
if (!(CVertical as Story).storyName) {
  (CVertical as Story).storyName = '1. 垂直间距';
  (CVertical as Story).parameters = {
    docs: {
      description: {
        story: '设置`direction`为`vertical`可以开启垂直模式。\n\n',
      },
    },
  };
}
if (!(CWrap as Story).storyName) {
  (CWrap as Story).storyName = '98. 环绕间距';
  (CWrap as Story).parameters = {
    docs: {
      description: {
        story: '设置`wrap`属性开启环绕模式。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Space/官网示例',
  tags: ['autodocs'],
};

export { CAlign, CBase, CCustomize, CSize, CSplit, CVertical, CWrap };
