// @universe-design/react/src/shared/components/picker/Comp/input-wrapper/index.tsx
import type { CSSProperties } from 'react';
import React from 'react';
import clsx from 'clsx';
import EmptyFilled from '@universe-design/icons-react/EmptyFilled';
import noop from 'lodash/noop';

export interface InputWrapperProps {
  style?: CSSProperties;
  className?: string;
  prefixCls?: string;
  suffixIcon?: React.ReactNode;
  allowClear?: boolean;
  showClear?: boolean;
  disabled?: boolean;
  focused?: boolean;
  onKeyDown?: React.KeyboardEventHandler<HTMLInputElement>;
  onMouseDown?: React.MouseEventHandler<HTMLDivElement>;
  onMouseUp?: React.MouseEventHandler<HTMLDivElement>;
  onClick?: React.MouseEventHandler<HTMLDivElement>;
  onClear?: () => void;
  children?: React.ReactNode;
}

function InputWrapper(props: InputWrapperProps) {
  const {
    // FIXME 为什么这个没有被用到？
    // style,
    className,
    prefixCls,
    allowClear = true,
    showClear,
    disabled,
    focused,
    onMouseDown,
    onMouseUp,
    onClick,
    onClear = noop,
    suffixIcon,
    children,
  } = props;

  const InputViewPrefixCls = `${prefixCls}-inputWrapper`;
  const isClearShow = allowClear && !disabled && showClear;

  return (
    <div
      onClick={onClick}
      onMouseDown={onMouseDown}
      onMouseUp={onMouseUp}
      className={clsx(className, {
        [InputViewPrefixCls]: true,
        [`${InputViewPrefixCls}__focused`]: focused,
        [`${InputViewPrefixCls}__disabled`]: disabled,
      })}
    >
      <div className={`${InputViewPrefixCls}-inner`}>
        {children}
        {suffixIcon && (
          <span
            className={clsx(`${InputViewPrefixCls}-icon`, isClearShow && `${InputViewPrefixCls}-suffixIcon__showClear`)}
          >
            {suffixIcon}
          </span>
        )}
        {isClearShow && (
          <span
            className={clsx(`${InputViewPrefixCls}-icon`, `${InputViewPrefixCls}-clear`)}
            onMouseDown={e => {
              e.preventDefault();
              e.stopPropagation();
            }}
            onMouseUp={e => {
              e.preventDefault();
              e.stopPropagation();
              onClear();
            }}
          >
            <EmptyFilled />
          </span>
        )}
      </div>
    </div>
  );
}

export default InputWrapper;
