// @universe-design/react/lib/react-mini-form/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Field", {
  enumerable: true,
  get: function () {
    return _Field.Field;
  }
});
Object.defineProperty(exports, "FieldPropsAreEqual", {
  enumerable: true,
  get: function () {
    return _Field.FieldPropsAreEqual;
  }
});
Object.defineProperty(exports, "Form", {
  enumerable: true,
  get: function () {
    return _Form.Form;
  }
});
Object.defineProperty(exports, "FormContextProvider", {
  enumerable: true,
  get: function () {
    return _FormContext.FormContextProvider;
  }
});
Object.defineProperty(exports, "FormList", {
  enumerable: true,
  get: function () {
    return _FormList.FormList;
  }
});
Object.defineProperty(exports, "FormSelector", {
  enumerable: true,
  get: function () {
    return _FormSelector.FormSelector;
  }
});
Object.defineProperty(exports, "FormSelectors", {
  enumerable: true,
  get: function () {
    return _FormSelector.FormSelectors;
  }
});
Object.defineProperty(exports, "marshalNamePath", {
  enumerable: true,
  get: function () {
    return _shared.marshalNamePath;
  }
});
Object.defineProperty(exports, "useField", {
  enumerable: true,
  get: function () {
    return _Field.useField;
  }
});
Object.defineProperty(exports, "useFieldContext", {
  enumerable: true,
  get: function () {
    return _useFieldInternal.useFieldContext;
  }
});
Object.defineProperty(exports, "useForm", {
  enumerable: true,
  get: function () {
    return _useFormInternal.useForm;
  }
});
Object.defineProperty(exports, "useFormContext", {
  enumerable: true,
  get: function () {
    return _FormContext.useFormContext;
  }
});
Object.defineProperty(exports, "useFormContextSelector", {
  enumerable: true,
  get: function () {
    return _FormContext.useFormContextSelector;
  }
});
Object.defineProperty(exports, "useFormList", {
  enumerable: true,
  get: function () {
    return _useFormList.useFormList;
  }
});
var _FormContext = require("./FormContext");
var _Form = require("./Form");
var _Field = require("./Field");
var _useFormInternal = require("./internal/useFormInternal");
var _useFieldInternal = require("./internal/useFieldInternal");
var _shared = require("./shared");
var _FormSelector = require("./FormSelector");
var _useFormList = require("./internal/useFormList");
var _FormList = require("./FormList");
