// @universe-design/react/lib/react-mini-form/shared.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DEFAULT_VALIDATE_TRIGGER = void 0;
exports.asArray = asArray;
exports.escapeNamePath = escapeNamePath;
exports.marshalNamePath = marshalNamePath;
exports.mergeWithUndefined = mergeWithUndefined;
exports.normalizeValidator = normalizeValidator;
exports.printFormValidationError = printFormValidationError;
exports.unescapeNamePath = unescapeNamePath;
exports.unmarshalNamePath = unmarshalNamePath;
var _isPlainObject = _interopRequireDefault(require("lodash/isPlainObject"));
var _util = require("../util");
/**
 * @public
 */

/**
 * @public
 */

/**
 * @public
 */

/**
 * @public
 */

/**
 * @public
 */

/**
 * @public
 */

const DEFAULT_VALIDATE_TRIGGER = exports.DEFAULT_VALIDATE_TRIGGER = 0x1f1f1f;
/**
 * @public
 */

/**
 * @public
 */

/**
 * @param value 被校验的字段值
 * @param serializedName 哪个字段触发
 * @param namePaths 触发哪些字段的校验
 * @param triggerName 触发事件
 *
 */

function asArray(data) {
  if (Array.isArray(data)) {
    return data;
  }
  if (data === undefined) {
    return [];
  }
  return [data];
}
function normalizeValidator(rawValidator) {
  if ((0, _util.isFunction)(rawValidator)) {
    return {
      validate: rawValidator,
      validateTrigger: asArray(rawValidator.validateTrigger)
    };
  }
  return {
    ...rawValidator,
    validateTrigger: asArray(rawValidator.validateTrigger)
  };
}
const SEP = '/';
const TILDE = '~';
const ESCAPED_TILDE = '~0';
const ESCAPED_SEP = '~1';
function escapeNamePath(name) {
  const path = String(name);
  if (path.indexOf(SEP) === -1 && path.indexOf(TILDE) === -1) {
    return path;
  }
  return path.replace(new RegExp(TILDE, 'g'), ESCAPED_TILDE).replace(new RegExp(SEP, 'g'), ESCAPED_SEP);
}
function unescapeNamePath(name) {
  return name.replace(new RegExp(ESCAPED_SEP, 'g'), SEP).replace(new RegExp(ESCAPED_TILDE, 'g'), TILDE);
}
function marshalNamePath(name) {
  return asArray(name).map(escapeNamePath).join(SEP);
}
function unmarshalNamePath(nameStr) {
  const parts = nameStr.split(SEP).map(unescapeNamePath);
  if (parts.length === 1 && parts[0] === '') {
    return [];
  }
  return parts;
}
function printFormValidationError(err) {
  console.warn('[ud: Form] Error occurs when processing validation. We use error message to determine dirty state of the form, please catch unexpected error and return or resolve error message', err);
}

/**
 * 解决 lodash merge 跳过 undefined 的问题, 以及 fields list 的 merge 逻辑问题
 * 1. 只要是 key in obj 对应的 value，我们都会进行 merge 处理。也就是说 a: undefined 会被执行 merge，而没有 a 这个 key 的时候就跳过 a 的 merge
 * 2. 所有的 array 都要进行判断:
 *    如果当前的 path 在 pathSet 中，则说明这个 array 是一个field，这时候的操作应该是 replace([1, 2, 3], [4, 5]) = [4, 5]
 *    如果当前的 path 不在 pathSet 中，则说明这个 array 表示一个 field list，这时候进行 index 对位 merge。merge([1, 2, 3], [4, 5]) = [4, 5, 3]
 */
function mergeWithUndefined(target, source, fieldPathSet) {
  let curPath = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : [];
  if (!source) {
    return target;
  }
  if (!target) {
    return source;
  }
  Object.keys(source).forEach(key => {
    const value = source[key];
    const prevValue = target[key];
    const nextPath = [...curPath, key];

    // 如果该 path 属于一个 field -> replace
    if (fieldPathSet.has(marshalNamePath(nextPath))) {
      target[key] = value;
    } else {
      // 如果 prevValue 不是 plain 的则可以直接覆盖，如果 value 不是 plain 的也可以直接覆盖
      const recursive = (Array.isArray(prevValue) || (0, _isPlainObject.default)(prevValue)) && (Array.isArray(value) || (0, _isPlainObject.default)(value));
      target[key] = recursive ? mergeWithUndefined(prevValue, value, fieldPathSet, nextPath) : value;
    }
  });
  return target;
}
