// @universe-design/react-input/lib-commonjs/hooks/useInput.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useInput = useInput;
var _isNil = _interopRequireDefault(require("lodash/isNil"));
var _reactUtilities = require("@universe-design/react-utilities");
var _useCursor = require("./useCursor");
var _useMaxLength = require("./useMaxLength");
var _react = require("react");
/**
 * 当用户传入不合法的值时，尝试将其专为 string
 */
function toString(rawValue) {
  if (typeof rawValue === 'string') {
    return rawValue;
  }
  if ((0, _isNil.default)(rawValue)) {
    return '';
  }
  return String(rawValue);
}
/**
 * 输入框的输入
 * @public
 */
function useInput(props) {
  const {
    value: valueProp,
    onChange,
    defaultValue,
    inputRef,
    slicer,
    counter: counterProp,
    truncateExcessWords,
    showCount,
    onCompositionStart: onCompositionStartProp,
    onCompositionEnd: onCompositionEndProp,
    maxLength: maxLengthProp,
    focused
  } = props;
  // 是否正在做输入合成
  const composing = (0, _react.useRef)(false);
  const [recordCursor, resetCursor] = (0, _useCursor.useCursor)(inputRef, focused);

  // 渲染后，重置光标位置
  (0, _react.useEffect)(() => {
    resetCursor();
  }, [resetCursor]);
  const {
    nativeMaxLength,
    counter,
    maxLength,
    valueFormatter: _valueFormatter
  } = (0, _useMaxLength.useMaxLength)({
    slicer,
    showCount,
    truncateExcessWords,
    counter: counterProp,
    maxLength: maxLengthProp
  });
  const valueFormatter = (0, _react.useCallback)(v => {
    // 当中文输入时，不能将其截断
    if (composing.current) {
      return v;
    }
    return _valueFormatter(v);
  }, [_valueFormatter]);
  const valueInProp = 'value' in props;
  const [value, setValue] = (0, _reactUtilities.useControlledState)(valueFormatter(toString(defaultValue)), valueFormatter(toString(valueProp)), {
    isControlledOverride: valueInProp
  });
  const onCompositionStart = (0, _react.useCallback)(e => {
    onCompositionStartProp === null || onCompositionStartProp === void 0 ? void 0 : onCompositionStartProp(e);
    composing.current = true;
  }, [onCompositionStartProp]);
  const onCompositionEnd = (0, _react.useCallback)(e => {
    composing.current = false;
    // 此时长度可能不正确，再 format 一次将其重置正确
    const formattedValue = valueFormatter(e.currentTarget.value);
    // 做 valueInProp 的判断是为了防止触发错误的渲染，导致无法输入中文
    if (!valueInProp) {
      setValue(formattedValue);
    }

    // 这里是 onChange 出格式化后的新值
    if (value !== formattedValue) {
      // onChange 前，先记录一下光标位置
      recordCursor();
      onChange === null || onChange === void 0 ? void 0 : onChange(e.currentTarget, e, formattedValue, 'composition');
    }
    // 要保证 compositionEnd 在 onChange 之后调用
    onCompositionEndProp === null || onCompositionEndProp === void 0 ? void 0 : onCompositionEndProp(e);
  }, [onChange, onCompositionEndProp, recordCursor, setValue, value, valueFormatter, valueInProp]);
  return {
    value,
    setValue,
    internalOnChange: e => {
      const formattedValue = valueFormatter(e.target.value);
      // 做 valueInProp 的判断是为了防止触发错误的渲染，导致无法输入中文
      if (!valueInProp) {
        setValue(formattedValue);
      }
      recordCursor();
      // TODO 这里可能会在 composition 结束的时候多调用一次未 truncate 的值，不过不影响使用，https://code.byted.org/universe-ui/react/issues/873#note_47790469
      onChange === null || onChange === void 0 ? void 0 : onChange(e.currentTarget, e, formattedValue, 'inputChange');
    },
    nativeMaxLength,
    count: counter(value),
    maxLength,
    valueFormatter,
    onCompositionStart,
    onCompositionEnd
  };
}
