// @universe-design/react-input/lib-commonjs/hooks/useMaxLength.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useMaxLength = useMaxLength;
/**
 * 最大长度的问题
 * 1. 默认使用按照码点计数的规则
 * 2. 支持用户传入自定义长度计算方法
 * 2. 支持用户传入自定义截断方法
 */

function useMaxLength(args) {
  const {
    showCount = false,
    maxLength,
    slicer: slicerProp,
    counter: counterProp,
    truncateExcessWords: truncateExcessWordsProp
  } = args;

  // 优先使用原生的 maxLength: 当没有设置 counter, slicer 相关属性时；
  const useNativeMaxLength = !(showCount || counterProp || slicerProp);
  // 是否截断
  const truncateExcessWords = maxLength !== undefined && (useNativeMaxLength || truncateExcessWordsProp);
  const counter = str => {
    if (counterProp) {
      return counterProp(str);
    }
    if (useNativeMaxLength) {
      return str.length;
    }
    return [...str].length;
  };
  return {
    valueFormatter: str => {
      if (!str) {
        return '';
      }
      // 如果未设置 maxLength 或者未超出长度，不截断
      if (!truncateExcessWords || counter(str) <= maxLength) {
        return str;
      }
      if (slicerProp) {
        return slicerProp(str, maxLength);
      }
      if (useNativeMaxLength) {
        return str.slice(0, maxLength);
      }
      return [...str].slice(0, maxLength).join('');
    },
    // 如果指定了 showCounter， 或者用户自己传入了 counter， 则认为不使用原生的 max
    nativeMaxLength: useNativeMaxLength ? maxLength : undefined,
    counter,
    maxLength
  };
}
