// @universe-design/react/lib/date-picker/cell/HoleMask.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HoleMask = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
/**
 * 用于扣出一个桥形，本来想用于画 date-range，但是最终效果不好，所以先提交在这里但是不使用
 * https://bytedance.feishu.cn/docx/doxcn9kqBZVTrbolR9ilVUzsR1X
 */
const HoleMask = _ref => {
  let {
    style,
    ...props
  } = _ref;
  return /*#__PURE__*/_react.default.createElement("svg", (0, _extends2.default)({
    height: "100%",
    width: "200%",
    xmlns: "http://www.w3.org/2000/svg",
    style: {
      position: 'absolute',
      fill: 'currentColor',
      ...style
    }
  }, props), /*#__PURE__*/_react.default.createElement("defs", null, /*#__PURE__*/_react.default.createElement("mask", {
    id: "hole"
  }, /*#__PURE__*/_react.default.createElement("rect", {
    width: "100%",
    height: "100%",
    fill: "#fff"
  }), /*#__PURE__*/_react.default.createElement("circle", {
    cx: "50%",
    cy: "50%",
    r: "50%",
    fill: "#000"
  }))), /*#__PURE__*/_react.default.createElement("rect", {
    width: "100%",
    height: "100%",
    mask: "url(#hole)"
  }));
};
exports.HoleMask = HoleMask;
