// @universe-design/react/lib/date-picker/cell/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useDateCellInteractiveAreaCls = exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _hooks = require("../../config-provider/hooks");
var _Range = require("./Range.class");
require("./index.less");
/**
 * DateCell 包含三个部分
 * - DateCellRangeBG，range 背景
 * - DateCellRangePreviewBG，range 预选背景
 * - DateCellInner，真正的日期元胞
 *
 * range 背景会有 start 和 end 的概念，会出现半圆形
 * range preview 和 range 是有叠加需求的
 */

const usePickerCellCssScope = () => (0, _hooks.useCssScope)('picker__cell');
const DateCell = /*#__PURE__*/_react.default.forwardRef((_ref, ref) => {
  let {
    children,
    isSelected = false,
    isDisabled = false,
    isToday = false,
    isOutOfView = false,
    isRangePreview = false,
    isRange = false,
    className,
    prevCellStatus,
    nextCellStatus,
    mode = 'cell',
    colIndex,
    colNum,
    ...restProps
  } = _ref;
  const cssScope = usePickerCellCssScope();
  const rangeCssScope = (0, _hooks.useCssScope)('picker__cell__range');
  if (mode === 'text') {
    return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
      ref: ref,
      className: (0, _clsx.default)(cssScope(), className)
    }, restProps), /*#__PURE__*/_react.default.createElement("div", {
      className: (0, _clsx.default)(cssScope('__inner'), cssScope('__inner-text'))
    }, children));
  }
  if (!prevCellStatus || !nextCellStatus || colIndex === undefined || colNum === undefined) {
    return null;
  }

  /**
   * 这段逻辑的解释在如下文档中
   * https://bytedance.feishu.cn/docx/doxcn9kqBZVTrbolR9ilVUzsR1X
   */

  // 全局截断

  const isRowEnd = colIndex === colNum - 1;
  const {
    cellColor: rangeCellColor,
    beforeColor: rangeBeforeColor,
    afterColor: rangeAfterColor
  } = new _Range.Range(prevCellStatus, {
    isSelected,
    isDisabled,
    isToday,
    isOutOfView,
    isRangePreview,
    isRange
  }, nextCellStatus);
  const rangeBeforeBGCls = (() => {
    if (colIndex === 0) {
      return;
    }
    return rangeBeforeColor ? rangeCssScope(`-bg-${rangeBeforeColor}`) : undefined;
  })();
  const rangeAfterBGCls = (() => {
    if (isRowEnd) {
      return;
    }
    return rangeAfterColor ? rangeCssScope(`-bg-${rangeAfterColor}`) : undefined;
  })();

  // Range 跟前面一段相接的部分
  const beforeRangeNode = /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(rangeCssScope('__before'), rangeBeforeBGCls)
  });

  // Range 跟后面一段相接的部分
  const afterRangeNode = /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(rangeCssScope('__after'), rangeAfterBGCls)
  });
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
    ref: ref,
    className: (0, _clsx.default)(cssScope(), isSelected && cssScope('-selected'), isOutOfView && cssScope('-outOfView'), isDisabled && cssScope('-disabled'), isToday && cssScope('-today'), isRange && cssScope('-range'), isRangePreview && cssScope('-rangePreview'), className)
  }, restProps), beforeRangeNode, afterRangeNode, /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(cssScope('__range'), rangeCellColor ? rangeCssScope(`-bg-${rangeCellColor}`) : undefined)
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: cssScope('__inner')
  }, children))), !isRowEnd && /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(cssScope('__space'), rangeAfterBGCls)
  }));
});
const useDateCellInteractiveAreaCls = () => usePickerCellCssScope()('-interactive-area');
exports.useDateCellInteractiveAreaCls = useDateCellInteractiveAreaCls;
var _default = exports.default = DateCell;
