// @universe-design/react/src/tag/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Basic from '../demo/0.basic';
import C1Closable from '../demo/1.closable';
import C2_0Attribute from '../demo/2.0.attribute';
import C2_1AttributeSolid from '../demo/2.1.attribute-solid';
import C2_2Option from '../demo/2.2.option';
import C2_3Status from '../demo/2.3.status';
import C2_4Avatar from '../demo/2.4.avatar';
import C2_5Special from '../demo/2.5.special';
import C2_6CustomizedColor from '../demo/2.6.customized-color';
import C3Size from '../demo/3.size';
import C4Icon from '../demo/4.icon';
import C5Shape from '../demo/5.shape';
import C6TagGroup from '../demo/6.tag-group';
if (!(C0Basic as Story).storyName) {
  (C0Basic as Story).storyName = '0. 基础用法';
  (C0Basic as Story).parameters = {
    docs: {
      description: {
        story: '标签的简单用法。\n\n',
      },
    },
  };
}
if (!(C1Closable as Story).storyName) {
  (C1Closable as Story).storyName = '1. 可关闭标签';
  (C1Closable as Story).parameters = {
    docs: {
      description: {
        story:
          '可以通过添加 `closable` 变为可关闭标签。可关闭标签具有 `onClose` 事件。也可通过 `visible` 属性控制标签的显示隐藏。\n\n',
      },
    },
  };
}
if (!(C2_0Attribute as Story).storyName) {
  (C2_0Attribute as Story).storyName = '2. 属性标签';
  (C2_0Attribute as Story).parameters = {
    docs: {
      description: {
        story:
          '属性标签用于描述事物、提供辅助信息，以帮助用户快速的识别。默认是属性标签带透明度，可以在更多背景下有良好的表现效果。\n\n',
      },
    },
  };
}
if (!(C2_1AttributeSolid as Story).storyName) {
  (C2_1AttributeSolid as Story).storyName = '2.1. 不透明属性标签';
  (C2_1AttributeSolid as Story).parameters = {
    docs: {
      description: {
        story:
          '每一种颜色可配置透明和不透明的标签底色。常规情况下使用带有透明度的底色。在背景为非常规界面的情况下（如非 bg，hover 等常规背景）可以使用不透明的标签底色。\n\n',
      },
    },
  };
}
if (!(C2_2Option as Story).storyName) {
  (C2_2Option as Story).storyName = '2.2. 选项标签';
  (C2_2Option as Story).parameters = {
    docs: {
      description: {
        story:
          '选项标签用于表示事物本身。标签颜色对比度较高，饱和度较低，不建议在系统层面建立标签颜色意义（如警示，正面，负面）。\n\n',
      },
    },
  };
}
if (!(C2_3Status as Story).storyName) {
  (C2_3Status as Story).storyName = '6. 状态标签';
  (C2_3Status as Story).parameters = {
    docs: {
      description: {
        story: '推荐的状态标签实现\n\n',
      },
    },
  };
}
if (!(C2_4Avatar as Story).storyName) {
  (C2_4Avatar as Story).storyName = '2.4. 头像标签';
  (C2_4Avatar as Story).parameters = {
    docs: {
      description: {
        story: '常用于展示人员、组织的名称、照片。\n\n',
      },
    },
  };
}
if (!(C2_5Special as Story).storyName) {
  (C2_5Special as Story).storyName = '2.5. 特殊标签示例';
  (C2_5Special as Story).parameters = {
    docs: {
      description: {
        story: '飞书业务中的特殊标签。\n\n',
      },
    },
  };
}
if (!(C2_6CustomizedColor as Story).storyName) {
  (C2_6CustomizedColor as Story).storyName = '2.6. 自定义颜色标签';
  (C2_6CustomizedColor as Story).parameters = {
    docs: {
      description: {
        story: '如果传入颜色不是预设值，则会认为是自定义颜色标签。\n\n',
      },
    },
  };
}
if (!(C3Size as Story).storyName) {
  (C3Size as Story).storyName = '3. 不同尺寸';
  (C3Size as Story).parameters = {
    docs: {
      description: {
        story: '标签具有中、小两种尺寸，可以在不同场景下选择合适的尺寸。\n\n',
      },
    },
  };
}
if (!(C4Icon as Story).storyName) {
  (C4Icon as Story).storyName = '4. 带图标的标签';
  (C4Icon as Story).parameters = {
    docs: {
      description: {
        story: '可通过设置 `icon` 在标签中加入图标。也可以直接将头像传入 `icon` 作为头像标签使用。\n\n',
      },
    },
  };
}
if (!(C5Shape as Story).storyName) {
  (C5Shape as Story).storyName = '5. 标签形状';
  (C5Shape as Story).parameters = {
    docs: {
      description: {
        story: '标签具有三种形状，`rect` 矩形、`round` 圆角矩形、`square` 正方形。\n\n',
      },
    },
  };
}
if (!(C6TagGroup as Story).storyName) {
  (C6TagGroup as Story).storyName = '6. 动态编辑标签';
  (C6TagGroup as Story).parameters = {
    docs: {
      description: {
        story: '可动态添加和删除标签。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Tag/官网示例',
  tags: ['autodocs'],
};

export {
  C0Basic,
  C1Closable,
  C2_0Attribute,
  C2_1AttributeSolid,
  C2_2Option,
  C2_3Status,
  C2_4Avatar,
  C2_5Special,
  C2_6CustomizedColor,
  C3Size,
  C4Icon,
  C5Shape,
  C6TagGroup,
};
