// @byted-service/metainfo/lib/consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrefixTransientUpstream = exports.PrefixTransient = exports.PrefixPersistent = exports.HTTPPrefixTransient = exports.HTTPPrefixPersistent = exports.persistentType = exports.transientType = exports.transientUpstreamType = exports.invalidType = exports.bwCtxKeyRecv = exports.bwCtxKeySend = exports.bwCtxKey = exports.ctxVersionKey = exports.ctxKey = exports.ctxVersion = exports.version = void 0;
// tslint:disable-next-line: no-var-requires
exports.version = require('../package.json').version;
exports.ctxVersion = 1631664000000; // 2021-09-15
exports.ctxKey = Symbol.for('K_META_INFO_CONTEXT');
exports.ctxVersionKey = Symbol.for('K_META_INFO_CONTEXT_VERSION');
exports.bwCtxKey = Symbol.for('K_META_INFO_BACKWARD');
exports.bwCtxKeySend = Symbol.for('K_META_INFO_BACKWARD_SEND');
exports.bwCtxKeyRecv = Symbol.for('K_META_INFO_BACKWARD_RECV');
exports.invalidType = 0;
exports.transientUpstreamType = 2;
exports.transientType = 4;
exports.persistentType = 8;
// HTTP header prefixes.
exports.HTTPPrefixPersistent = 'rpc-persist-';
exports.HTTPPrefixTransient = 'rpc-transit-';
// The prefix listed below may be used to tag the types of values when there is no context to carry them.
exports.PrefixPersistent = 'RPC_PERSIST_';
exports.PrefixTransient = 'RPC_TRANSIT_';
exports.PrefixTransientUpstream = 'RPC_TRANSIT_UPSTREAM_';
