// @byted-nodex/metrics/lib/fields.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Fields = void 0;
exports.newCounterField = newCounterField;
exports.newRateCounterField = newRateCounterField;
exports.newTimerField = newTimerField;
exports.newMeterField = newMeterField;
exports.newStoreField = newStoreField;
exports.newDeltaCounterField = newDeltaCounterField;
exports.newHistogramField = newHistogramField;
const metrics_codec_1 = require("@byted-nodex/metrics-codec");
const consts_1 = require("./consts");
const assert_1 = __importDefault(require("assert"));
function newCodecField(name) {
    return {
        name,
        type: metrics_codec_1.FieldType.Float64,
    };
}
function buildTimerCodecFields(suffix) {
    return consts_1.defaultTimerFieldsSuffixes.map(f => newCodecField(`${suffix ? `${suffix}.` : ''}${f}`));
}
function buildDeltaCounterFields(suffix) {
    return consts_1.defaultDeltaSuffixes.map(f => newCodecField(`${suffix ? `${suffix}.` : ''}${f}`));
}
function buildHistogramFields(buckets, suffix) {
    // hist:xxx:b-0
    const fields = buckets.map(v => {
        return newCodecField(`hist:${suffix ? `${suffix}:` : ''}b-${v}`);
    });
    // hist:xxx:sum / hist:xxx:count
    consts_1.defaultHistoSuffixes.forEach(v => {
        fields.push(newCodecField(`hist:${suffix ? `${suffix}:` : ''}${v}`));
    });
    return fields;
}
const defaultTimerFields = buildTimerCodecFields();
const defaultDeltaFields = buildDeltaCounterFields();
class Fields {
    constructor(fields, withSuffixAsFieldName) {
        // 排重
        this.originFields = fields.reverse().filter((f, index, arr) => {
            return arr.findIndex(ff => ff.name === f.name && ff.type === f.type) === index;
        });
        this.fields = new Map();
        this.expand(withSuffixAsFieldName);
    }
    set(field) {
        this.fields.set(`${field.type}.${field.name}`, field);
    }
    get(type, name) {
        return this.fields.get(`${type}.${name}`);
    }
    checkBuckets(buckets) {
        (0, assert_1.default)(buckets && buckets.length, 'buckets is invalid, must have at least one bucket');
        if (buckets.length > consts_1.bucketMaxCount) {
            throw new Error(`too many buckets: ${buckets.length}, exceeded limit ${consts_1.bucketMaxCount}`);
        }
    }
    getDefaultTimerFields() {
        return defaultTimerFields;
    }
    getDefaultDeltaFields() {
        return defaultDeltaFields;
    }
    getDefaultHistogramFields(field) {
        var _a, _b;
        // histogram 的 fields 跟 buckets 关联，所以没有默认 fields
        return (_b = (_a = this.get(metrics_codec_1.MetricType.Histogram, field)) === null || _a === void 0 ? void 0 : _a.meta) === null || _b === void 0 ? void 0 : _b.codecFields;
    }
    expand(withSuffixAsFieldName) {
        const fields = this.originFields.slice(0);
        fields.forEach(f => {
            const field = Object.create(f);
            const meta = (field.meta = {
                codecFields: [],
            });
            if (field.type === metrics_codec_1.MetricType.Timer) {
                meta.codecFields = buildTimerCodecFields(withSuffixAsFieldName ? field.name : undefined);
            }
            else if (field.type === metrics_codec_1.MetricType.DeltaCounterType) {
                meta.codecFields = buildDeltaCounterFields(withSuffixAsFieldName ? field.name : undefined);
            }
            else if (field.type === metrics_codec_1.MetricType.Histogram) {
                this.checkBuckets(field.buckets);
                meta.buckets = field.buckets;
                meta.codecFields = buildHistogramFields(meta.buckets, withSuffixAsFieldName ? field.name : undefined);
            }
            else {
                meta.codecFields.push(newCodecField(field.name));
                if (field.type === metrics_codec_1.MetricType.Meter) {
                    // 改成 counter
                    field.type = metrics_codec_1.MetricType.Counter;
                    // 要多一个 .rate 的 rateCounter
                    const rateName = `${field.name}.rate`;
                    const rateField = newRateCounterField(rateName);
                    rateField.meta = {
                        codecFields: [newCodecField(rateName)],
                    };
                    this.set(rateField);
                }
            }
            this.set(field);
        });
    }
}
exports.Fields = Fields;
function newCounterField(name) {
    return { type: metrics_codec_1.MetricType.Counter, name };
}
function newRateCounterField(name) {
    return { type: metrics_codec_1.MetricType.RateCounter, name };
}
function newTimerField(name) {
    return { type: metrics_codec_1.MetricType.Timer, name };
}
function newMeterField(name) {
    return { type: metrics_codec_1.MetricType.Meter, name };
}
function newStoreField(name) {
    return { type: metrics_codec_1.MetricType.Store, name };
}
function newDeltaCounterField(name) {
    return { type: metrics_codec_1.MetricType.DeltaCounterType, name };
}
function newHistogramField(name, buckets) {
    return {
        type: metrics_codec_1.MetricType.Histogram,
        name,
        // buckets 必须从小到大排序
        buckets: buckets.sort((a, b) => a - b),
    };
}
