// @byted-nodex/metrics/lib/measure.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Measure = void 0;
exports.measure = measure;
exports.incr = incr;
exports.observe = observe;
exports.store = store;
exports.incrMeter = incrMeter;
exports.incrCounter = incrCounter;
exports.stat = stat;
exports.add = add;
exports.withField = withField;
exports.withSuffix = withSuffix;
const metrics_codec_1 = require("@byted-nodex/metrics-codec");
const consts_1 = require("./consts");
function measure(metricType, value, suffix) {
    return {
        type: metricType,
        suffix: typeof suffix === 'string' ? suffix : consts_1.defaultMeasureSuffix[metricType],
        value,
    };
}
class Measure {
    constructor(opt) {
        this.suffix = opt.suffix;
    }
    /** Add increments an int value as the delta_counter type, It has three fields: delta, rate, and counter. */
    add(value) {
        return measure(metrics_codec_1.MetricType.DeltaCounterType, value, this.suffix);
    }
    /** Incr increments a rate counter type metric as a default suffix "rate". */
    incr(value) {
        return measure(metrics_codec_1.MetricType.RateCounter, value, this.suffix);
    }
    /** Observe increments a timer type metric as a default suffix "timer". */
    observe(value) {
        return measure(metrics_codec_1.MetricType.Timer, value, this.suffix);
    }
    /** Store store a metric as a default suffix "store". */
    store(value) {
        return measure(metrics_codec_1.MetricType.Store, value, this.suffix);
    }
    /** IncrMeter increments a meter type metric as a default suffix "meter". */
    incrMeter(value) {
        return measure(metrics_codec_1.MetricType.Meter, value, this.suffix);
    }
    /** IncrCounter increments a counter type metric as a default suffix "counter". */
    incrCounter(value) {
        return measure(metrics_codec_1.MetricType.Counter, value, this.suffix);
    }
    /** Stat increments a histogram type metric as a default suffix "histogram". */
    stat(value) {
        return measure(metrics_codec_1.MetricType.Histogram, value, this.suffix);
    }
}
exports.Measure = Measure;
/** Incr increments a rate counter type metric as a default suffix "rate". */
function incr(value) {
    return measure(metrics_codec_1.MetricType.RateCounter, value);
}
/** Observe increments a timer type metric as a default suffix "timer". */
function observe(value) {
    return measure(metrics_codec_1.MetricType.Timer, value);
}
/** Store store a metric as a default suffix "store". */
function store(value) {
    return measure(metrics_codec_1.MetricType.Store, value);
}
/** IncrMeter increments a meter type metric as a default suffix "meter". */
function incrMeter(value) {
    return measure(metrics_codec_1.MetricType.Meter, value);
}
/** IncrCounter increments a counter type metric as a default suffix "counter". */
function incrCounter(value) {
    return measure(metrics_codec_1.MetricType.Counter, value);
}
/** Stat increments a histogram type metric as a default suffix "histogram". */
function stat(value) {
    return measure(metrics_codec_1.MetricType.Histogram, value);
}
/** Add increments an int value as the delta_counter type, It has three fields: delta, rate, and counter. */
function add(value) {
    return measure(metrics_codec_1.MetricType.DeltaCounterType, value);
}
/**
 * withField 跟 withSuffix 逻辑其实一致，都是添加指标后缀，逻辑跟 go sdk 一致
 * 比如最终上报的完整指标名是 toutiao.nodejs.sdk.counter ，那么其中 counter 就是指标后缀，但是多值不太一样：
 * 如果是单值，那么上报到 agent 的指标就是完整的 `toutiao.nodejs.sdk.counter`
 * 如果是多值，那么上报到 agent 的指标就是 `toutiao.nodejs.sdk` ，然后 field 是 counter
 */
function withField(field) {
    return new Measure({ suffix: field });
}
function withSuffix(suffix) {
    return new Measure({ suffix });
}
