// @byted-nodex/metrics/lib/util.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isValidTagValue = exports.isValidString = exports.silentLogger = void 0;
exports.noop = noop;
exports.isAgentSidecar = isAgentSidecar;
exports.getSeqpacketSocketPath = getSeqpacketSocketPath;
exports.formatTagValue = formatTagValue;
exports.percentile = percentile;
exports.sortTagNames = sortTagNames;
exports.uniq = uniq;
exports.simpleHash = simpleHash;
exports.getSdkVersion = getSdkVersion;
exports.getTime = getTime;
exports.serializeMetricOption = serializeMetricOption;
exports.genUniqId = genUniqId;
exports.isNil = isNil;
exports.getHost = getHost;
exports.isBytedHostname = isBytedHostname;
exports.hostIpToHost = hostIpToHost;
exports.getDefaultTags = getDefaultTags;
exports.getIDC = getIDC;
exports.readIDCFromMetricsEdgeConf = readIDCFromMetricsEdgeConf;
exports.overwriteDCForPPE = overwriteDCForPPE;
exports.getPPEFlag = getPPEFlag;
exports.getHostEnv = getHostEnv;
exports.readPPEFlagFromFile = readPPEFlagFromFile;
exports.timeoutPromise = timeoutPromise;
const consts_1 = require("./consts");
const ip_1 = __importDefault(require("ip"));
const fs_1 = __importDefault(require("fs"));
const os_1 = __importDefault(require("os"));
const env_1 = __importDefault(require("@byted-service/env"));
const path_1 = __importDefault(require("path"));
let sdkVersion = '1.0.0';
try {
    const pkgInfo = JSON.parse(fs_1.default.readFileSync(path_1.default.resolve(__dirname, '../package.json')).toString());
    sdkVersion = pkgInfo.version;
}
catch (_a) {
    // ignore
}
const numberChars = '0123456789';
const lowerLetterChars = 'abcdefghijklmnopqrstuvwxyz';
const commonChars = (numberChars + lowerLetterChars + lowerLetterChars.toUpperCase()).split('');
const whiteListNameCharCodes = [...commonChars, '.', '_', '-', '/', ':', '%'].map(v => v.charCodeAt(0));
const whiteListValueCharCodes = [
    ...commonChars,
    ' ',
    '!',
    '#',
    '$',
    '%',
    '&',
    "'",
    '+',
    ',',
    '-',
    '.',
    '/',
    ':',
    ';',
    '<',
    '=',
    '>',
    '?',
    '@',
    '[',
    '\\',
    ']',
    '^',
    '_',
    '`',
    '~',
].map(v => v.charCodeAt(0));
const whiteListNameChars = [];
const whiteListValueChars = [];
for (let i = 0; i <= consts_1.maxASCIICode; i++) {
    whiteListNameChars.push(whiteListNameCharCodes.includes(i) ? 1 : 0);
    whiteListValueChars.push(whiteListValueCharCodes.includes(i) ? 1 : 0);
}
function noop(..._args) {
    // do nothing
}
exports.silentLogger = {
    error: noop,
    info: noop,
    warn: noop,
};
function isAgentSidecar() {
    return process.env.METRICSERVER2_SIDECAR === '1';
}
function getSeqpacketSocketPath() {
    if (isAgentSidecar()) {
        return '/sidecar/sock/metric_seqpacket.sock';
    }
    else {
        return '/opt/tmp/sock/metric_seqpacket.sock';
    }
}
const isValidString = (s) => {
    if (s.length === 0 || s.length > consts_1.maxTagLen) {
        return false;
    }
    for (let i = 0; i < s.length; i++) {
        const r = s.charCodeAt(i);
        if (r > consts_1.maxASCIICode) {
            return false;
        }
        if (whiteListNameChars[r] !== 1) {
            return false;
        }
    }
    return true;
};
exports.isValidString = isValidString;
const isValidTagValue = (s) => {
    if (s.length === 0 || s.length > consts_1.maxTagLen) {
        return false;
    }
    // 不允许空格开头或者结尾
    if (s[0] === ' ' || s[s.length - 1] === ' ') {
        return false;
    }
    for (let i = 0; i < s.length; i++) {
        const r = s.charCodeAt(i);
        if (r > consts_1.maxASCIICode) {
            return false;
        }
        if (whiteListValueChars[r] !== 1) {
            return false;
        }
    }
    return true;
};
exports.isValidTagValue = isValidTagValue;
function formatTagValue(tagValue) {
    if (isNil(tagValue)) {
        return '-';
    }
    tagValue = String(tagValue);
    if (!(0, exports.isValidTagValue)(tagValue)) {
        return '-';
    }
    return tagValue;
}
function percentile(s, percent) {
    return s[Math.floor(s.length * percent)];
}
function sortTagNames(tagNames) {
    return tagNames.sort((a, b) => (a < b ? -1 : a > b ? 1 : 0));
}
function uniq(list) {
    return Array.from(new Set(list));
}
const offset32 = 2166136261;
function simpleHash(ss) {
    let h = offset32;
    for (const v of ss) {
        for (let i = 0; i < v.length; i++) {
            // simple hash alg but faster than fnv32a
            // https://gist.github.com/hyamamoto/fd435505d29ebfa3d9716fd2be8d42f0
            h = (h << 5) - h + v.charCodeAt(i);
            h |= 0; // Convert to 32bit integer
        }
    }
    return Math.abs(h);
}
function getSdkVersion() {
    return sdkVersion;
}
function getTime(now, interval) {
    const mod = now % interval;
    if (mod === 0) {
        return [now, 0];
    }
    else {
        const nextTime = Math.ceil(now / interval) * interval;
        const duration = nextTime - now;
        return [nextTime, duration];
    }
}
function serializeMetricOption(options) {
    let key = options.name +
        '|' +
        (options.tenant || '-') +
        '|' +
        (options.compatTimer ? '1' : '0') +
        '|' +
        (options.forceMultiFields ? '1' : '0') +
        '|' +
        (options.keepAlive ? '1' : '0');
    if (options.tagNames) {
        for (const name of options.tagNames) {
            key += '|' + name;
        }
    }
    return key;
}
let idx = 0;
function genUniqId() {
    return idx++;
}
function isNil(v) {
    return v === undefined || v === null;
}
function getHost() {
    const hostname = process.env.HOSTNAME || os_1.default.hostname() || '';
    if (isBytedHostname(hostname)) {
        return hostname;
    }
    return hostIpToHost(env_1.default.getHostIP());
}
function isBytedHostname(hostname) {
    // n112-028-209-194
    // n148-050-105
    // p23-080-123
    // dc03-pff-500-28c6-641b-f93-8de2
    // dc03-p14-tc2c-n052
    return hostname.startsWith('n') || hostname.startsWith('p') || hostname.startsWith('dc');
}
function padIp(n) {
    return n.padStart(3, '0');
}
function hostIpToHost(ip) {
    if (ip.includes('.')) {
        if (ip.startsWith('10.')) {
            const ips = ip.split('.');
            return `n${ips[1]}-${padIp(ips[2])}-${padIp(ips[3])}`;
        }
        if (ip.startsWith('33.')) {
            const ips = ip.split('.');
            return `p${ips[1]}-${padIp(ips[2])}-${padIp(ips[3])}`;
        }
        const ips = ip.split('.');
        if (ips.length === 4) {
            return `n${padIp(ips[0])}-${padIp(ips[1])}-${padIp(ips[2])}-${padIp(ips[3])}`;
        }
        return ip;
    }
    else if (ip.includes(':')) {
        // ipv6
        const ele = ip.split(':');
        if (!ele.length || ele[0] !== 'fdbd' || ele.length < 5) {
            return ip;
        }
        if (ele.filter(e => e !== '').length === 5) {
            return `${ele[1]}-p${ele[2]}-t${ele[3]}-n${ele[ele.length - 1].padStart(3, '0')}`;
        }
        const sb = [];
        const buf = ip_1.default.toBuffer(ip, Buffer.alloc(16));
        for (let i = 2; i < 16; i += 2) {
            let str = '';
            if (i == 4) {
                str += 'p';
            }
            if (buf[i] !== 0) {
                str += buf[i].toString(16);
                str += buf[i + 1].toString(16).padStart(2, '0');
            }
            else {
                str += buf[i + 1].toString(16);
            }
            sb.push(str);
        }
        return sb.join('-');
    }
    return '-';
}
/** 环境 tags */
function getDefaultTags() {
    const defaultTags = {
        dc: getIDC(),
        host: getHost(),
        _pid: process.pid,
        _env: env_1.default.getEnv(),
        _psm: env_1.default.getPSM(),
        _cluster: env_1.default.getCluster(),
        _ipv4: env_1.default.getIPv4(),
        _ipv6: env_1.default.getIPv6(),
    };
    if (env_1.default.isInTCE()) {
        const isSidecar = process.env.TCE_IS_INJECTED === 'true' ? 1 : 0;
        return {
            ...defaultTags,
            _pod_ip: process.env.MY_POD_IP || process.env.MY_POD_IPv6 || '-',
            _pod_name: env_1.default.getPodName(),
            _tce_physical_cluster: process.env.TCE_PHYSICAL_CLUSTER || '-',
            _deploy_stage: env_1.default.getStage(),
            _env_type: process.env.TCE_HOST_ENV || '-',
            _image_version: env_1.default.getImageVersion(),
            _is_sidecar: process.env.TCE_IS_INJECTED === 'true' ? '1' : '0',
            _primary_psm: isSidecar === 1 ? process.env.TCE_PRIMARY_PSM_IN_SIDECAR : undefined,
        };
    }
    return defaultTags;
}
function getIDC() {
    if (process.env.METRICS_SANDBOX) {
        return 'fake_sandbox';
    }
    // priority: BYTED_INJECT_DC env > metrics_edge_conf > consul_idc
    if (process.env.BYTED_INJECT_DC) {
        return process.env.BYTED_INJECT_DC;
    }
    let idc = readIDCFromMetricsEdgeConf(consts_1.kMetricsEdgeConfFile);
    if (!idc) {
        idc = env_1.default.getIDC();
    }
    const newIDC = overwriteDCForPPE(idc);
    return newIDC;
}
function readIDCFromMetricsEdgeConf(file) {
    try {
        if (!fs_1.default.existsSync(file)) {
            return '';
        }
        const content = fs_1.default.readFileSync(file, 'utf8');
        let reg = /^EDGE_MACHINE=edge-([^-\s]+)-([^-\s]+)/;
        let result = reg.exec(content);
        if (result === null || result === void 0 ? void 0 : result[2]) {
            return result[2];
        }
        reg = /^EDGE_MACHINE=external_edge-([^-\s]+)-([^-\s]+)/;
        result = reg.exec(content);
        if (result === null || result === void 0 ? void 0 : result[2]) {
            return result[2];
        }
        reg = /^EDGE_MACHINE=internal_edge-([^-\s]+)-([^-\s]+)/;
        result = reg.exec(content);
        if (result === null || result === void 0 ? void 0 : result[2]) {
            return result[2];
        }
        return '';
    }
    catch (_a) {
        return '';
    }
}
function overwriteDCForPPE(idc) {
    const ppeFlag = getPPEFlag();
    switch (ppeFlag) {
        case 'PPE':
        case 'ppe':
            switch (idc) {
                case 'lf':
                case 'hl':
                case 'lq':
                case 'yg':
                case 'lfzg':
                case 'hlzg':
                    return 'ppe';
                case 'va':
                case 'maliva':
                    return 'ppe-va';
                case 'sg':
                case 'sg1':
                case 'alisg':
                case 'sgsaas1larkidc1':
                case 'sgsaas1larkidc2':
                case 'sgsaas1larkidc3':
                    return 'ppe-sig';
                case 'sg2':
                    return 'ppe-sig2';
                case 'useast2a':
                    return 'ppe-useast2a';
                case 'useast5':
                    return 'ppe-useast5';
                default:
                    return 'ppe-' + idc;
            }
        case 'ppe-va':
            return 'ppe-va';
        case 'ppe-sig':
            return 'ppe-sig';
    }
    return idc;
}
function getPPEFlag() {
    if (getHostEnv().toLowerCase().includes('ppe')) {
        return 'ppe';
    }
    if (fs_1.default.existsSync(consts_1.kOsPPEFile)) {
        return 'ppe';
    }
    if (process.env.TCE_PHYSICAL_CLUSTER) {
        const str = process.env.TCE_PHYSICAL_CLUSTER.toLowerCase();
        if (str.includes('ppe')) {
            return str;
        }
    }
    let ppeFlag = readPPEFlagFromFile(consts_1.kTceClusterFile);
    if (ppeFlag) {
        return ppeFlag;
    }
    ppeFlag = readPPEFlagFromFile(consts_1.kTceSubClusterFile);
    if (ppeFlag) {
        return ppeFlag;
    }
    return '';
}
function getHostEnv() {
    if (process.env.TCE_HOST_ENV) {
        return process.env.TCE_HOST_ENV;
    }
    return '-';
}
function readPPEFlagFromFile(file) {
    let fd;
    try {
        if (!fs_1.default.existsSync(file)) {
            return '';
        }
        fd = fs_1.default.openSync(file, 'r');
        const buf = Buffer.alloc(4096);
        fs_1.default.readSync(fd, buf, { offset: 0, length: 4096 });
        const lines = buf.toString().split('\n');
        for (const line of lines) {
            if (line.toLowerCase().includes('ppe')) {
                return line;
            }
        }
        return '';
    }
    catch (_a) {
        return '';
    }
    finally {
        if (fd) {
            fs_1.default.closeSync(fd);
        }
    }
}
function timeoutPromise(p, timeout, errMsg) {
    return Promise.race([
        new Promise(p),
        new Promise((_, reject) => {
            setTimeout(() => {
                reject(new Error(errMsg));
            }, timeout);
        }),
    ]);
}
