// @universe-design/react/src/pin-input/PinInputHyphen.tsx
import stylex from '@universe-design/stylex';
import clsx from 'clsx';
import React from 'react';
import { sxEmptyElement } from '../stylex';
import type { StylexProps } from '../util';
import { forwardHTMLRef } from '../util/forwardRef';
import { sxPinInputMargin } from './sx';
import { usePinInputContext } from './usePinInput';

const sxPinInputHyphen = {
  root: stylex.create({
    display: 'inline-flex',
    alignItems: 'baseline',
    justifyContent: 'center',
  }),
  content: stylex.create({
    alignSelf: 'center',
    flexShrink: 0,
    width: 8,
    height: 1,
    transform: 'translateY(1px)',
  }),
};

export const PinInputHyphen = forwardHTMLRef<StylexProps, 'div'>(({ sx, ...props }, ref) => {
  const { size, rootCssScope } = usePinInputContext();
  return (
    <div
      ref={ref}
      {...props}
      className={clsx(
        rootCssScope('-hyphen'),
        stylex(sxPinInputHyphen.root, sxPinInputMargin[size], ...(sx ?? [])),
        props.className,
      )}
    >
      <div className={clsx(rootCssScope('-hyphen-content'), stylex(sxPinInputHyphen.content))} />
      <div className={stylex(sxEmptyElement)}>{'\xa0'}</div>
    </div>
  );
});
