// @universe-design/react/src/pin-input/usePinInput.tsx
import React, { useImperativeHandle } from 'react';
import stylex from '@universe-design/stylex';
import clsx from 'clsx';
import { useEffectOnce } from '../hooks';
import type { DescendantsManager } from '@universe-design/react-utilities';
import {
  createContextHook,
  KeyboardKey,
  useControlledState,
  createDescendantContext,
  useConstant,
} from '@universe-design/react-utilities';
import type { DefaultSizeType } from '../util/size';
import { getSizeAbbr } from '../util/size';
import type { ExtendedFocusOptions, StylexProps } from '../util';
import { focus, isNonNullable } from '../util';
import { composeEventHandlers } from '../util/function';
import composeRefs from '../util/composeRefs';
import { useCssScope } from '../config-provider';
import type { NativeInputProps } from '../shared/components/Input';
import { sxBorderRadius, sxInput, sxInputWidth, sxTransitionTimingFunction } from '../stylex';
import { sxPinInputMargin } from './sx';

import './style';

/**
 * ================================
 *
 * Create context to track descendants and their indices
 *
 * ================================
 */
export const [
  PinInputDescendantsProvider,
  usePinInputDescendantsContext,
  usePinInputDescendants,
  usePinInputDescendant,
] = createDescendantContext<HTMLInputElement>();

/**
 * ================================
 *
 * Create Context
 *
 * ================================
 */
export type PinInputContext = Omit<UsePinInputReturn, 'descendants'>;

export const { Provider: PinInputContextProvider, useContext: usePinInputContext } = createContextHook<PinInputContext>(
  'PinInputContext',
  {
    strict: true,
  },
);

/**
 * ================================
 *
 * Pin Input
 *
 * ================================
 */

/**
 * @public
 */
export interface UsePinInputProps {
  /**
   * 子元素
   */
  children: React.ReactNode;
  /**
   * 是否自动聚焦
   * @defaultValue false
   */
  autoFocus?: boolean;
  /**
   * 输入内容的受控值
   */
  value?: string;
  /**
   * 输入内容的默认值
   */
  defaultValue?: string;
  /**
   * 尺寸大小
   * @defaultValue `medium`
   */
  size?: DefaultSizeType;
  /**
   * 输入框内容变化时触发的回调
   */
  onChange?: (value: string) => void;
  /**
   * 输入框内容变化，当所有输入值都有效时的回调
   */
  onComplete?: (value: string) => void;
  /**
   * 占位符
   */
  placeholder?: string;
  /**
   * id 属性
   * @example
   * 如果 `id='foo'`，第一个 input 的 id 为 `foo-0`
   */
  id?: string;
  /**
   * 组件是否为禁用状态
   * @defaultValue false
   */
  disabled?: boolean;
  /**
   * 组件是否为错误状态
   * @defaultValue false
   */
  error?: boolean;
  /**
   * 是否为加密状态，类似于 type = 'password'
   * @defaultValue false
   */
  mask?: boolean;
  /**
   * @defaultValue `number`
   * 允许输入的字符集。它们对应的字符集为：数字、字母、数字和字母
   */
  type?: 'number' | 'alphabetic' | 'alphanumeric';
}

export interface PinInputRef {
  /**
   * 清空 pin-input
   */
  clear: () => void;
  /**
   *
   */
  focus: HTMLInputElement['focus'];
}

const toCharArray = (value: string) => value.split('');
const validate = (v: string, type: UsePinInputProps['type']) => {
  let regex = /^[0-9]+$/;

  if (type === 'alphanumeric') {
    regex = /^[0-9a-zA-Z]+$/;
  }
  if (type === 'alphabetic') {
    regex = /^[a-zA-Z]+$/;
  }
  return regex.test(v);
};
/**
 * 用于判断当前应该在哪个 input 中输入，规则是第一个空值的 input
 */
const getShouldInputIndex = (values: string[], descendants: DescendantsManager<any, any>) => {
  for (let i = 0; i < values.length; i++) {
    if (!values[i]) {
      return i;
    }
  }
  return Math.min(values.length, descendants.count() - 1);
};

export const usePinInput = (props: UsePinInputProps, ref: React.ForwardedRef<PinInputRef>) => {
  const {
    autoFocus = false,
    value,
    defaultValue = '',
    size = 'medium',
    onChange,
    onComplete,
    placeholder = '',
    id,
    disabled = false,
    error = false,
    mask = false,
    type = 'number',
  } = props;

  const rootCssScope = useCssScope('pin-input');
  const descendants = usePinInputDescendants();

  const [focusedIndex, setFocusedIndex] = React.useState(-1);

  const [values, setValues] = useControlledState(
    // defaultValue 需要去除所有的非法字符
    toCharArray(defaultValue).filter(v => validate(v, type)) || [],
    // value
    isNonNullable(value) ? toCharArray(value).filter(v => validate(v, type)) : [],
    // options
    {
      onChange: React.useCallback(
        (nextValues: string[]) => {
          onChange?.(nextValues.join(''));
          // complete 放在 change 后面
          const isComplete = nextValues.every(Boolean) && nextValues.length === descendants.count();
          if (isComplete) {
            onComplete?.(nextValues.join(''));
          }
        },
        [descendants, onChange, onComplete],
      ),
      isControlledOverride: 'value' in props,
    },
  );

  /**
   * 设置下一个 input 为聚焦状态
   */
  const focusNext = React.useCallback(
    (index: number) => {
      const next = descendants.next(index, false);
      if (next) {
        focus(next.node, { selectTextIfInput: false });
      }
    },
    [descendants],
  );

  const focusByIndex = React.useCallback(
    (index: number, options?: ExtendedFocusOptions) => {
      const descendant = descendants.values()[index];
      if (descendant) {
        focus(descendant.node, { selectTextIfInput: false, ...options });
      }
    },
    [descendants],
  );

  /**
   * 设置 input 的单字符 value
   */
  const setValue = React.useCallback(
    (newValue: string, index: number) => {
      const nextValues = [...values];
      nextValues[index] = newValue;
      focusNext(index);
      setValues(nextValues);
    },
    [values, focusNext, setValues],
  );

  /**
   * 重置 pin-input 输入框内容，并 focus 第一个 input
   */
  const clear = React.useCallback(() => {
    const nextValues: string[] = Array(descendants.count()).fill('');
    setValues(nextValues);
    const first = descendants.first();
    if (first) {
      focus(first.node, { selectTextIfInput: false });
    }
  }, [descendants, setValues]);

  /**
   * autoFocus 逻辑，只生效一次，后续的变化都不触发
   */
  useEffectOnce(() => {
    if (autoFocus) {
      // useEffect 是比 useLayoutEffect 里使用 requestAnimation 和 setTimeout 调用时机都要早
      focusByIndex(getShouldInputIndex(values, descendants));
    }
  });

  useImperativeHandle(
    ref,
    () => ({
      clear,
      focus: option => {
        focusByIndex(getShouldInputIndex(values, descendants), option);
      },
    }),
    [clear, descendants, focusByIndex, values],
  );

  return {
    descendants,
    rootCssScope,
    // props
    id,
    size,
    placeholder,
    disabled,
    error,
    mask,
    // states
    values,
    focusedIndex,
    // actions
    setValue,
    setValues,
    clear,
    setFocusedIndex,
    focusByIndex,
    type,
  };
};

export type UsePinInputReturn = ReturnType<typeof usePinInput>;

/**
 * ================================
 *
 * Pin Input Field
 *
 * ================================
 */

/**
 * @public
 */
export type UsePinInputFieldProps = NativeInputProps & StylexProps;

export type PinInputFieldRef = HTMLInputElement;

const sxPinInputField = {
  root: stylex.compose(
    stylex.create({
      boxSizing: 'border-box',
      fontSize: 14,
      textAlign: 'center',
      borderWidth: 1,
      borderStyle: 'solid',
      ':focus-visible': {
        outline: 'none',
      },
      transitionProperty: 'border-color',
      transitionDuration: '0.1s',
    }),
    sxBorderRadius.sm,
    sxTransitionTimingFunction.standard,
  ),
  small: stylex.compose(sxInput.small, sxInputWidth.small),
  medium: stylex.compose(sxInput.medium, sxInputWidth.medium),
  large: stylex.compose(sxInput.large, sxInputWidth.large),
  complete: stylex.create({
    caretColor: 'transparent',
  }),
};

export const usePinInputField = (
  { sx, ...props }: UsePinInputFieldProps,
  ref: React.ForwardedRef<PinInputFieldRef>,
) => {
  const {
    setValue,
    setValues,
    focusByIndex,
    setFocusedIndex,
    focusedIndex,
    mask,
    disabled,
    id,
    values,
    placeholder,
    size,
    rootCssScope,
    error,
    type,
  } = usePinInputContext();
  const { index, descendants, register } = usePinInputDescendant();
  const inputRef = React.useRef<HTMLInputElement>(null);
  const latestValues = useConstant(() => ({
    get index() {
      // https://github.com/chakra-ui/chakra-ui/blob/main/packages/descendant/src/use-descendant.ts#L59
      return inputRef.current?.dataset.index ? Number(inputRef.current.dataset.index) : -1;
    },
  }));

  const hasFocus = focusedIndex === index;
  const isActive = focusedIndex >= 0 && focusedIndex >= index;
  const isComplete = Boolean(index === descendants.count() - 1 && values[index]);

  const inputProps: NativeInputProps & React.RefAttributes<HTMLInputElement> = {
    'aria-label': 'Please enter your pin code',
    inputMode: 'numeric',
    type: mask ? 'password' : 'tel',
    id: isNonNullable(id) ? `${id}-${index}` : undefined,
    disabled,
    value: values[index] || '',
    placeholder: hasFocus ? '' : placeholder,
    ref: composeRefs(ref, inputRef, register),
    autoComplete: 'off',
    tabIndex: getShouldInputIndex(values, descendants) === index ? undefined : -1, // TODO 由于 descendant register 不会触发其他参与项目的 rerender。这个值在 input field 动态变化的时候不一定准，例如，一个 4 个 field 当前 focus 到 index 3，此时，在不改变 focusIndex 的情况下修改 field 为 3 个，这时就完全无法 tab 到 pin-input 上
    ...props,
    className: clsx(
      rootCssScope('-field'),
      rootCssScope(`-field-${getSizeAbbr(size)}`),
      isActive && rootCssScope('-field-active'),
      error && rootCssScope('-field-error'),
      disabled && rootCssScope('-field-disabled'),
      stylex(
        ...(sx ?? [
          sxPinInputField.root,
          sxPinInputField[size],
          isComplete && sxPinInputField.complete,
          index !== 0 && sxPinInputMargin[size],
        ]),
      ),
      props.className,
    ),
    style: {
      ...(mask
        ? {
            fontFamily: 'Tahoma, PingFang SC, Microsoft Yahei', // https://codesandbox.io/s/infallible-feather-j4oez3?file=/src/App.tsx 在测试中发现 tahoma 是 win 和 mac 都系统自带的效果最好的字体，剩下两个是各自系统默认
          }
        : {}),
      ...props.style,
    },
    onChange: composeEventHandlers(props.onChange, event => {
      const eventValue = event.currentTarget.value;
      // input 中的 value 被删除
      if (eventValue === '') {
        setValue('', index);
        return;
      }

      // 输入值不合法
      if (!validate(eventValue, type)) {
        return;
      }

      // 普通输入元素的时候
      if (eventValue.length === 1) {
        setValue(eventValue, index);
        return;
      }

      // 如果是最后一个元素，输入 2 个字符时需要覆盖当前值
      if (index === descendants.count() - 1) {
        setValue(eventValue[1], index);
        return;
      }

      // 复制粘贴的时候
      const nextValues = values
        .slice(0, index)
        .concat(eventValue.split(''))
        .filter((_, i) => i < descendants.count());
      focusByIndex(nextValues.length - 1);
      setValues(nextValues);
    }),
    onKeyDown: composeEventHandlers(props.onKeyDown, event => {
      if (event.key === KeyboardKey.Backspace) {
        if (event.currentTarget.value === '') {
          const preInput = descendants.prev(index, false);
          if (preInput) {
            setValue('', index - 1);
            focus(preInput.node, { selectTextIfInput: false });
          }
        } else {
          setValue('', index);
        }
      }
    }),
    onMouseDown: composeEventHandlers(props.onMouseDown, ev => {
      // 如果当前 input 不该被 focus 就 preventDefault
      const shouldInputIndex = getShouldInputIndex(values, descendants);
      if (shouldInputIndex !== index) {
        ev.preventDefault();
        focusByIndex(shouldInputIndex);
      }
    }),
    onFocusChange: (focused, ev) => {
      props.onFocusChange?.(focused, ev);
      if (focused) {
        // autoFocus 的时候 index 还没有更新，但是 dom 上的 index 已经更新了，所以需要从 dom 上取 index。
        setFocusedIndex(latestValues.index);
      }
    },
    onBlur: composeEventHandlers(props.onBlur, () => {
      setFocusedIndex(-1);
    }),
  };

  return { inputProps };
};
