// @byted-service/metrics/lib/metrics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Metrics = void 0;
const util_1 = require("util");
const protocol_1 = require("./protocol");
const pool_1 = require("./pool");
const flusher_1 = require("./flusher");
const auth_1 = require("./auth");
const query_1 = require("./query");
const util_2 = require("./util");
const consts_1 = require("./consts");
// NOTE 先设置成 500
const MAX_AGG_SIZE = 500;
const MAX_METRIC_AGG_SIZE = 5000;
const MAX_TAG_SIZE = 500;
const METRIC_TYPE_NAME_SEPARATOR = '___$_$___';
const METRICS_ENABLE = process.env.METRICS_ENABLE !== 'false';
const flusher = new flusher_1.Flusher();
class Metrics {
    static createMetrics(options) {
        return new Metrics(options);
    }
    constructor(options = {}) {
        this._buffer = [];
        this._preDefinedTagKeys = new Map();
        // Map<`${metricType}${METRIC_TYPE_NAME_SEPARATOR}${metricName}`,Map<tagKeyValue,value>>
        this._aggregationCaches = new Map();
        this.debug =
            typeof options.debug === 'boolean'
                ? options.debug
                : consts_1.defaultOptions.debug;
        const silentVerifying = typeof options.silentVerifying === 'boolean'
            ? options.silentVerifying
            : consts_1.defaultOptions.silentVerifying;
        this.host = options.host || consts_1.defaultOptions.host;
        this.port = options.port || consts_1.defaultOptions.port;
        this.defaultPrefix =
            typeof options.defaultPrefix === 'string'
                ? options.defaultPrefix
                : consts_1.defaultOptions.defaultPrefix;
        this.bufferSize =
            typeof options.bufferSize === 'number'
                ? options.bufferSize
                : consts_1.defaultOptions.bufferSize;
        this.flushInterval =
            typeof options.flushInterval === 'number'
                ? options.flushInterval
                : consts_1.defaultOptions.flushInterval;
        this.logger = options.logger || consts_1.defaultOptions.logger;
        this.enablePreAggregate =
            typeof options.enablePreAggregate === 'boolean'
                ? options.enablePreAggregate
                : consts_1.defaultOptions.enablePreAggregate;
        this._send = this._send || this._udpSend;
        const disableBuiltinClusterTag = options.disableBuiltinClusterTag ||
            consts_1.defaultOptions.disableBuiltinClusterTag;
        this.tceTagsStr = (0, util_2.getTceTagsStr)('|', disableBuiltinClusterTag);
        const fns = [
            ['emitCounter', 'counter'],
            ['emitTimer', 'timer'],
            ['emitStore', 'store'],
            ['emitRateCounter', 'rate_counter'],
            ['emitMeter', 'meter'],
        ];
        for (const [key, type] of fns) {
            this[key] = function (name, value, prefix, tags, tagKeys) {
                name = `${prefix || this.defaultPrefix}.${name}`;
                if (this.enablePreAggregate &&
                    tags &&
                    this._preDefinedTagKeys.size > 0) {
                    const key = `${type}${METRIC_TYPE_NAME_SEPARATOR}${name}`;
                    const preDefinedTagKeys = this._preDefinedTagKeys.get(key);
                    // 如果没有预定义 tag key 直接走兜底逻辑
                    if (preDefinedTagKeys) {
                        let aggCache = this._aggregationCaches.get(key);
                        if (!aggCache &&
                            this._aggregationCaches.size < MAX_AGG_SIZE) {
                            aggCache = new Map();
                            this._aggregationCaches.set(key, aggCache);
                        }
                        // 没有聚合缓存直接走兜底逻辑
                        // tagKey-value 的组合类型总数超过 MAX_METRIC_AGG_SIZE 限制后不再做聚合
                        // 这里一共会有 500 * 5000 = 250 万种组合的可能
                        if (aggCache && aggCache.size < MAX_METRIC_AGG_SIZE) {
                            const tagsStr = (0, util_2.toKV)(tags, preDefinedTagKeys);
                            // store 类型直接覆盖
                            if (type === 'store') {
                                aggCache.set(tagsStr, value);
                            }
                            else {
                                aggCache.set(tagsStr, (aggCache.get(tagsStr) || 0) + value);
                            }
                            flusher.register(this.flushInterval, this);
                            return;
                        }
                    }
                }
                const tagsStr = (0, util_2.toKV)(tags, tagKeys);
                this._buffer.push([
                    type,
                    name,
                    String(value),
                    tagsStr ? `${this.tceTagsStr}|${tagsStr}` : this.tceTagsStr,
                ]);
                flusher.register(this.flushInterval, this);
                if (this._buffer.length >= this.bufferSize) {
                    this.flush().catch(consts_1.noop);
                }
            };
        }
        if (!silentVerifying) {
            for (const [key] of fns) {
                const fn = this[key];
                this[key] = function (name, value, prefix, tags, tagKeys) {
                    if (!name || typeof name !== 'string') {
                        throw new Error((0, util_1.format)('invalid metrics %s, name is invalid', `${prefix || this.defaultPrefix}.${name}`));
                    }
                    if (!value && value !== 0) {
                        throw new Error((0, util_1.format)('invalid metrics %s, value is invalid, value: %s', `${prefix || this.defaultPrefix}.${name}`, value));
                    }
                    if (tags) {
                        if (typeof tags !== 'object' || Array.isArray(tags)) {
                            throw new Error((0, util_1.format)('invalid metrics %s, tags is invalid', `${prefix || this.defaultPrefix}.${name}`));
                        }
                        const keys = tagKeys || Object.keys(tags);
                        for (let i = 0; i < keys.length; i++) {
                            const key = keys[i];
                            const val = String(tags[key]);
                            // 规则参考: https://bytedance.feishu.cn/docs/GHFmzle2R6a7cGvAqMWlbc#CQBkoH
                            if (!(0, util_2.isValidString)(key)) {
                                this.logger.warn('[metrics] invalid metrics %s, tag "%s" is invalid. See https://bytedance.feishu.cn/docs/GHFmzle2R6a7cGvAqMWlbc#CQBkoH', `${prefix || this.defaultPrefix}.${name}`, key);
                            }
                            if (!(0, util_2.isValidString)(val)) {
                                this.logger.warn('[metrics] invalid metrics %s, tag %s\'s value "%s" is invalid. See https://bytedance.feishu.cn/docs/GHFmzle2R6a7cGvAqMWlbc#CQBkoH', `${prefix || this.defaultPrefix}.${name}`, key, val);
                            }
                        }
                    }
                    return fn.call(this, name, value, prefix, tags, tagKeys);
                };
            }
        }
        // 用户主动关闭 Metrics 发送，打印提示
        if (!METRICS_ENABLE) {
            const warningMessage = 'Will NOT send metrics to server due to environment variables(METRICS_ENABLE=false).';
            if (this.debug) {
                // 调试状态输出到 STDOUT
                process.stdout.write(`\x1b[36mMETRIC\x1b[0m ${warningMessage.replace('NOT', '\x1B[31mNOT\x1b[0m')}\n`);
            }
            else {
                this.logger.warn(warningMessage);
            }
            for (const [key] of fns) {
                this[key] = consts_1.noop;
            }
        }
        if (this.debug) {
            for (const [key] of fns) {
                const fn = this[key];
                this[key] = function (name, value, prefix, tags, tagKeys) {
                    process.stdout.write(`\x1b[36mMETRIC\x1b[0m ${[
                        `${prefix || this.defaultPrefix}.${name}`,
                        value,
                        (0, util_2.toKV)(tags, tagKeys),
                    ].join(' ')}\n`);
                    return fn.call(this, name, value, prefix, tags, tagKeys);
                };
            }
        }
    }
    /**
     * 预声明 tag key
     */
    preDefineTagKey(options) {
        if (!this.enablePreAggregate) {
            return;
        }
        // 存储 key
        const { keys, metricType, name, prefix } = options;
        const metricName = `${prefix || this.defaultPrefix}.${name}`;
        const typeMetricName = `${metricType}${METRIC_TYPE_NAME_SEPARATOR}${metricName}`;
        const preDefinedTagKeys = this._preDefinedTagKeys.get(typeMetricName) || [];
        for (let i = 0; i < keys.length; i++) {
            const key = keys[i];
            if (preDefinedTagKeys.length >= MAX_TAG_SIZE) {
                this.logger.warn('[metrics] Can not append more, reach max predefined tag key size: %d', MAX_TAG_SIZE);
                break;
            }
            if (!(0, util_2.isValidString)(key)) {
                this.logger.warn('[metrics] Can not prefine tag key "%s"', key);
                continue;
            }
            if (!preDefinedTagKeys.includes(key)) {
                preDefinedTagKeys.push(key);
            }
        }
        this._preDefinedTagKeys.set(typeMetricName, preDefinedTagKeys);
    }
    async flush() {
        if (this._aggregationCaches.size > 0) {
            for (const [key, aggCache] of this._aggregationCaches) {
                const [type, name] = key.split(METRIC_TYPE_NAME_SEPARATOR);
                for (const [tag, value] of aggCache) {
                    this._buffer.push([
                        type,
                        name,
                        String(value),
                        `${this.tceTagsStr}|${tag}`,
                    ]);
                }
                aggCache.clear();
            }
        }
        const metrics = this._buffer;
        this._buffer = [];
        if (metrics.length > 0) {
            await this._send(metrics);
        }
    }
    /**
     * 开发者不需要使用metrics后应该要调用close
     */
    close() {
        this.flush().catch(consts_1.noop);
    }
    /** @deprecated */
    reset() {
        this.flush().catch(consts_1.noop);
    }
    async _udpSend(metrics) {
        const packets = (0, protocol_1.encode)(metrics, consts_1.maxPacketSize, this.logger);
        const socket = await pool_1.pool.acquire(this);
        for (let i = 0; i < packets.length; i++) {
            const packet = packets[i];
            try {
                socket.send(packet, 0, packet.byteLength, this.port, this.host);
            }
            catch (err) {
                //
            }
        }
        pool_1.pool.release(socket);
    }
}
exports.Metrics = Metrics;
Metrics.Auth = auth_1.Auth;
Metrics.query = query_1.query;
Metrics.parseTimeString = query_1.parseTimeString;
Metrics.kHost = util_2.kHost;
Metrics.kUsHost = util_2.kUsHost;
Metrics.getAuthPrefix = util_2.getAuthPrefix;
Metrics.getQueryPrefix = util_2.getQueryPrefix;
