// @universe-design/react/src/text/Text.tsx
/** @jsxRuntime classic */
/** @jsx jsx */
import * as React from 'react';
import stylex from '@universe-design/stylex';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx, useMergeComponentProps } from '@universe-design/stylex/react';
import {
  cssTypographyBody0,
  cssTypographyBody2,
  cssTypographyCaption0,
  cssTypographyCaption1,
  cssTypographyCaption1Hack,
  cssTypographyHeadline,
  cssTypographyTitle0,
  cssTypographyTitle1,
  cssTypographyTitle2,
  cssTypographyTitle3,
  cssTypographyTitle4,
  cssTypographyTitle5,
  sxLineClamp,
  sxOverflowEllipsis,
} from '../stylex';
import { genericForwardRef } from '../util/forwardRef';
import { useTextOverflowPopperSidecar } from '../config-provider/hooks/useTextOverflowPopperSidecar';
import { isHorizontalOverflowed, isVerticalOverflowed } from '../util/dom';
import { useCssScope } from '../config-provider';
import type { StylexProps } from '@universe-design/stylex/react';
import { isNonNullable } from '../util';
import type { TextOverflowPopperSidecarProps } from '../config-provider/interface';

export const fontStyles = {
  title0: cssTypographyTitle0,
  title1: cssTypographyTitle1,
  title2: cssTypographyTitle2,
  title3: cssTypographyTitle3,
  title4: cssTypographyTitle4,
  title5: cssTypographyTitle5,
  headline: cssTypographyHeadline,
  caption0: cssTypographyCaption0,
  caption1: cssTypographyCaption1,
  caption1Hack: cssTypographyCaption1Hack,
  body0: cssTypographyBody0,
  body2: cssTypographyBody2,
};

/**
 * @docgen
 */
export interface TextExoticProps extends StylexProps {
  /**
   * 排版样式类型，若设置为 `plain` 则排版样式会继承父级参数。
   * @defaultValue `body0`
   */
  variant?: keyof typeof fontStyles | 'plain';
  /**
   * 最多显示几行文本，若不传值则不会限制行数
   */
  lineClamp?: number;
  /**
   * 是否为禁用态，禁用态用户无法选中文字，文字溢出可以弹出提示
   * @defaultValue false
   */
  disabled?: boolean;
  /**
   * 文字溢出弹层组件属性
   */
  tooltipProps?: TextOverflowPopperSidecarProps;
  /**
   * 文字对齐方式
   */
  alignment?: 'start' | 'center' | 'end' | 'justify';
  /**
   * 是否使用斜体
   * @defaultValue false
   */
  italic?: boolean;
}

export const Text = genericForwardRef<
  TextExoticProps,
  'span',
  'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'p' | 'pre' | 'span' | 'div' | 'label'
>(
  (
    {
      as: C = 'span' as const,
      variant = 'body0',
      lineClamp,
      disabled = false,
      tooltipProps,
      alignment,
      italic = false,
      ...props
    },
    ref,
  ) => {
    /**
     * ====================
     *   Props
     * ====================
     */
    const { children } = props;
    const noClamp = !isNonNullable(lineClamp);
    const oneLineClamp = !noClamp && lineClamp <= 1;
    const multilineClamp = !oneLineClamp && !noClamp;

    /**
     * ====================
     *   Context
     * ====================
     */
    const cssScope = useCssScope('text');
    const TextOverflowPopper = useTextOverflowPopperSidecar();

    /**
     * ====================
     *   Styles
     * ====================
     */
    const cProps = useMergeComponentProps(props, {
      className: cssScope(),
      sx: React.useMemo(
        () => [
          stylex.create({ wordWrap: 'break-word', overflowWrap: 'break-word' }),
          variant !== 'plain' && fontStyles[variant],
          // 如果 span 并且开了 one line clamp 则需要使用 inline-block
          C === 'span' && oneLineClamp && stylex.create({ display: 'inline-block' }),
          // line clamp 里面有 display: -webkit-box 要放在其他的 display 后面
          (() => {
            if (oneLineClamp) {
              return sxOverflowEllipsis;
            }
            if (multilineClamp) {
              return sxLineClamp;
            }
          })(),
          // 默认从父级继承或使用 initial 值
          alignment &&
            {
              start: stylex.create({ textAlign: 'start' }),
              center: stylex.create({ textAlign: 'center' }),
              end: stylex.create({ textAlign: 'end' }),
              justify: stylex.create({ textAlign: 'justify' }),
            }[alignment],
          italic && stylex.create({ fontStyle: 'italic' }),
          disabled &&
            stylex.create({
              color: $token['udtoken-component-text-disabled-loading'],
              cursor: 'not-allowed',
              userSelect: 'none',
            }),
        ],
        [C, alignment, disabled, italic, multilineClamp, oneLineClamp, variant],
      ),
      style: {
        WebkitLineClamp: multilineClamp ? lineClamp : undefined,
      },
    });

    /**
     * ====================
     *   Render
     * ====================
     */
    const node = <C ref={ref} {...cProps} />;

    if (noClamp) {
      return node;
    }

    const checkOverflow = oneLineClamp ? isHorizontalOverflowed : isVerticalOverflowed;

    return (
      <TextOverflowPopper
        {...{
          title: children,
          checkOverflow,
          ...tooltipProps,
        }}
      >
        {node}
      </TextOverflowPopper>
    );
  },
);

/**
 * @docgen
 */
export type TextProps = React.ComponentProps<typeof Text>;

if (__DEV__) {
  Text.displayName = 'Text';
}
