// @universe-design/react/src/popconfirm/index.tsx
import type { ReactNode } from 'react';
import React, { forwardRef, useCallback, useEffect, useRef } from 'react';
import clsx from 'clsx';
import stylex from '@universe-design/stylex';
import type { BaseTooltipProps } from '../tooltip';
import { Tooltip } from '../tooltip';
import { useCssScope, useLocale } from '../config-provider/hooks';
import type { ButtonProps } from '../button';
import { Button } from '../button';
import InfoFilled from '@universe-design/icons-react/InfoFilled';
import SucceedFilled from '@universe-design/icons-react/SucceedFilled';
import WarningFilled from '@universe-design/icons-react/WarningFilled';
import ErrorFilled from '@universe-design/icons-react/ErrorFilled';
import type { LiteralUnion } from '../util';
import { isFunction, KeyCode } from '../util';
import type { RenderFunction } from '../util/getRenderPropValue';
import { getRenderPropValue } from '../util/getRenderPropValue';
import { cssBlockBase, cssTypographyHeadline } from '../stylex';
import type { PopperRef } from '../shared/components/popper';

import './style';

export type PopconfirmIconType = 'success' | 'warning' | 'info' | 'error';
export interface PopconfirmProps extends BaseTooltipProps {
  /**
   * 确认框的标题
   */
  title?: ReactNode | RenderFunction;
  /**
   * 确认框的描述
   */
  content?: ReactNode | RenderFunction;
  /**
   * 确认按钮文字
   * @defaultValue `确认`
   */
  okText?: string;
  /**
   * 取消按钮文字
   * @defaultValue `取消`
   */
  cancelText?: string;
  /**
   * 确定按钮的属性
   */
  okButtonProps?: ButtonProps;
  /**
   * 取消按钮的属性
   */
  cancelButtonProps?: ButtonProps;
  /**
   * 自定义弹出气泡图标
   * @defaultValue `info`
   */
  icon?: LiteralUnion<PopconfirmIconType, ReactNode> | null;
  /**
   * 点击取消的回调
   */
  onCancel?: (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;
  /**
   * 点击确认的回调
   */
  onConfirm?: (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;
}

const iconMap = {
  success: SucceedFilled,
  warning: WarningFilled,
  info: InfoFilled,
  error: ErrorFilled,
};

// eslint-disable-next-line max-lines-per-function
export const Popconfirm = forwardRef<PopperRef, PopconfirmProps>((props, ref) => {
  const {
    title,
    content,
    icon = 'info',
    trigger = 'click',
    okText,
    cancelText,
    okButtonProps = {},
    cancelButtonProps = {},
    onCancel,
    onConfirm,
    disabled = false,
    mouseLeaveDelay = 100,
    ...restProps
  } = props;

  const locale = useLocale();
  const cssScope = useCssScope('popconfirm');
  const overlayCssScope = cssScope('__overlay');
  const udTriggerRef = useRef<PopperRef>(null);

  const renderIcon = () => {
    if (!icon) {
      return null;
    }

    if (typeof icon === 'string' && (icon === 'success' || icon === 'warning' || icon === 'info' || icon === 'error')) {
      const Icon = iconMap[icon as PopconfirmIconType];
      return (
        <div className={cssScope(`-icon-${icon as PopconfirmIconType}`)}>
          <Icon />
        </div>
      );
    }

    return (
      <div className={cssScope('-icon-customer')}>
        <span className={cssScope('-icon-customer-wrapper')}>{icon}</span>
      </div>
    );
  };

  const handleOnConfirm = (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
    isFunction(okButtonProps.onClick) && okButtonProps.onClick(e);
    isFunction(onConfirm) && onConfirm(e);
    if (!e.isDefaultPrevented() && !disabled) {
      udTriggerRef.current?.setVisible(false);
    }
  };

  const handleOnCancel = (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
    isFunction(cancelButtonProps.onClick) && cancelButtonProps.onClick(e);
    isFunction(onCancel) && onCancel(e);
    if (!e.isDefaultPrevented() && !disabled) {
      udTriggerRef.current?.setVisible(false);
    }
  };

  const handleKeyDown = useCallback(
    (e: KeyboardEvent) => {
      if (e.keyCode === KeyCode.ESC && !e.defaultPrevented && !disabled) {
        udTriggerRef.current?.setVisible(false);
      }
    },
    [disabled],
  );

  useEffect(() => {
    window.addEventListener('keydown', handleKeyDown);
    return () => {
      window.removeEventListener('keydown', handleKeyDown);
    };
  }, [handleKeyDown]);

  const renderButton = () => (
    <div className={`${overlayCssScope}-button-group`}>
      <Button size="small" {...cancelButtonProps} onClick={handleOnCancel}>
        {cancelText || locale.components.popconfirm.cancelText}
      </Button>
      <Button style={{ marginRight: 0 }} type="primary" size="small" {...okButtonProps} onClick={handleOnConfirm}>
        {okText || locale.components.popconfirm.okText}
      </Button>
    </div>
  );

  const renderOverlay = () => {
    if (!title && !content) {
      return null;
    }
    return (
      <div className={overlayCssScope}>
        {renderIcon()}
        <div>
          {title && (
            <div className={clsx(`${overlayCssScope}-title`, stylex(cssBlockBase, cssTypographyHeadline))}>
              {getRenderPropValue(title)}
            </div>
          )}
          {content && <div className={`${overlayCssScope}-content`}>{getRenderPropValue(content)}</div>}
          {renderButton()}
        </div>
      </div>
    );
  };

  return (
    <Tooltip
      ref={ref}
      udTriggerRef={udTriggerRef}
      prefixCls={cssScope()}
      trigger={trigger}
      overlay={renderOverlay()}
      disabled={disabled}
      mouseLeaveDelay={mouseLeaveDelay}
      arrowHasStroke
      {...restProps}
    />
  );
});

export type { RenderFunction };
