// @universe-design/react/es/tooltip/Tooltip.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React, { cloneElement, forwardRef, isValidElement, useMemo } from 'react';
import clsx from 'clsx';
import { useCssScope } from '../config-provider/hooks';
import { usePopperSidecar } from '../config-provider/hooks/usePopperSidecar';
import './style';
export const splitObject = (obj, keys) => {
  const picked = {};
  const omitted = {
    ...obj
  };
  keys.forEach(key => {
    if (obj && key in obj) {
      picked[key] = obj[key];
      delete omitted[key];
    }
  });
  return {
    picked,
    omitted
  };
};

/**
 * @public
 * @docgen
 */

export const Tooltip = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    prefixCls,
    title,
    overlay,
    trigger = 'hover',
    // maxWidth, // max-width 现在加在了 popup 上可以直接用 overlayStyle 去覆盖
    placement = 'top',
    overlayStyle,
    overlayClassName,
    hideArrow: hideArrowProp = false,
    showArrow: showArrowProp = true,
    // adjustOverflow = true,
    children,
    onOverlayMouseEnter,
    onOverlayMouseLeave,
    mouseLeaveDelay = 0,
    wrapDisabledChildren = false,
    arrowHasStroke = false,
    ...restProps
  } = props;
  const cssScope = useCssScope('tooltip', prefixCls);
  const Popper = usePopperSidecar();
  const showArrow = 'hideArrow' in props ? !hideArrowProp : showArrowProp;
  const getDisabledCompatibleChildren = React.useCallback(element => {
    if (!element.props.disabled) {
      return element;
    }
    // Pick some layout related style properties up to span
    // Prevent layout bugs
    const {
      picked,
      omitted
    } = splitObject(element.props.style, ['position', 'left', 'right', 'top', 'bottom', 'float', 'display', 'zIndex']);
    const spanStyle = {
      display: 'inline-block',
      // default inline-block is important
      ...picked,
      cursor: 'not-allowed',
      width: element.props.block ? '100%' : null
    };
    const buttonStyle = {
      ...omitted,
      pointerEvents: 'none'
    };
    const child = /*#__PURE__*/cloneElement(element, {
      style: buttonStyle,
      className: null
    });
    return /*#__PURE__*/React.createElement("span", {
      style: spanStyle,
      className: clsx(element.props.className, `${cssScope()}-disabled-compatible-wrapper`)
    }, child);
  }, [cssScope]);
  const mergedChildren = useMemo(() => {
    if (! /*#__PURE__*/isValidElement(children)) {
      return /*#__PURE__*/React.createElement("span", {
        className: cssScope('__invalid-children-wrapper')
      }, children);
    }

    /**
     * if children is disabled Button or native disabled button,
     * wrap children with <span>,
     * because disabled button won't fire mouseenter/mouseleave/mouseover events for some browsers
     */
    if (wrapDisabledChildren) {
      return getDisabledCompatibleChildren(children);
    }
    return children;
  }, [children, cssScope, getDisabledCompatibleChildren, wrapDisabledChildren]);
  const popup = React.useMemo(() => /*#__PURE__*/React.createElement("div", {
    onMouseEnter: onOverlayMouseEnter,
    onMouseLeave: onOverlayMouseLeave,
    className: cssScope('-content'),
    role: "tooltip"
  }, title !== null && title !== void 0 ? title : overlay), [cssScope, onOverlayMouseEnter, onOverlayMouseLeave, overlay, title]);
  if (!(title !== null && title !== void 0 ? title : overlay)) {
    return /*#__PURE__*/React.createElement(React.Fragment, null, children);
  }
  return /*#__PURE__*/React.createElement(Popper, _extends({
    ref: ref,
    prefixCls: cssScope(),
    popupPlacement: placement,
    action: trigger,
    popup,
    children: mergedChildren,
    // arrow
    showArrow,
    arrowHasStroke,
    // popupMotion
    popupMotion: 'tooltip-zoom',
    popupStyle: overlayStyle,
    popupClassName: overlayClassName,
    mouseLeaveDelay,
    defaultMainAxisOffsetWithoutArrow: 4
  }, restProps, {
    __screenshotAlwaysOpenEnabled: true
  }));
});
if (process.env.NODE_ENV !== "production") {
  Tooltip.displayName = 'Tooltip';
}
