// @universe-design/react/src/shared/components/expand-button/index.tsx
import React from 'react';
import { ExpandableExpandDownFilled } from './ExpandDownFilled';
import { ExpandableDownBoldOutlined } from './DownBoldOutlined';
import clsx from 'clsx';
import type { ButtonProps, ButtonRefType } from '../../../button';
import { Button } from '../../../button';
import { useCssScope } from '../../../config-provider/hooks';
import { isFunction } from '../../../util';

import './index.less';

export interface ExpandButtonProps extends Omit<ButtonProps, 'children' | 'icon' | 'type' | 'size'> {
  /**
   * 是否为展开状态
   */
  isExpanded: boolean;
  onExpandChange?: (isExpanded: boolean) => void;
  icon?: React.ReactNode | ((props: { isExpanded: boolean }) => React.ReactNode);
  /**
   * 按钮大小
   */
  size?: ButtonProps['size'] | 'xxs';
  /**
   * 使用什么类型的展开图标，分别对应不同尺寸场景
   */
  iconType?: 'ExpandDownFilled' | 'DownBoldOutlined';
  /**
   * 是否把展开图标作为占位符使用，可以在不展示展开按钮时保证父子元素对齐
   */
  asPlaceholder?: boolean;
  /**
   * 展开图标是否仅做 UI 展示，禁用 hover、active、点击事件
   */
  readonly?: boolean;
}

export const ExpandButton = React.forwardRef<ButtonRefType, ExpandButtonProps>(
  (
    {
      isExpanded,
      icon,
      iconType,
      size = iconType === 'ExpandDownFilled' ? 'xxs' : 'small',
      onExpandChange,
      onClick,
      className,
      asPlaceholder = false,
      readonly = false,
      ...restProps
    },
    ref,
  ) => {
    const cssScope = useCssScope('expandButton');
    const handleClick = React.useCallback<React.MouseEventHandler<HTMLElement>>(
      e => {
        isFunction(onClick) && !readonly && onClick(e);
        isFunction(onExpandChange) && onExpandChange(!isExpanded);
      },
      [isExpanded, onClick, onExpandChange, readonly],
    );

    const iconNode = (() => {
      if (isFunction(icon)) {
        return icon({ isExpanded });
      }
      if (typeof icon !== 'undefined') {
        return icon;
      }
      if (iconType === 'ExpandDownFilled') {
        return <ExpandableExpandDownFilled isExpanded={isExpanded} />;
      }
      if (iconType === 'DownBoldOutlined') {
        return <ExpandableDownBoldOutlined isExpanded={isExpanded} />;
      }
    })();

    return (
      <Button
        type="icon"
        ref={ref}
        className={clsx(
          cssScope(),
          className,
          asPlaceholder && cssScope('-as-placeholder'),
          readonly && cssScope('-readonly'),
          size === 'xxs' && cssScope('-xxs'),
        )}
        icon={iconNode}
        onClick={handleClick}
        size={size === 'xxs' ? undefined : size}
        {...restProps}
      />
    );
  },
);
