// @universe-design/styles/esm/im/index.dark.atom.js
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
export const im_btn_selected_bg_gray_hover = 'rgba(235, 235, 235, 0.12)';
export const im_btn_selected_bg_gray_normal = 'rgba(235, 235, 235, 0.08)';
export const im_btn_selected_bg_gray_press = 'rgba(235, 235, 235, 0.16)';
export const im_btn_selected_bg_green_hover = 'rgba(65, 158, 52, 0.28)';
export const im_btn_selected_bg_green_normal = 'rgba(65, 158, 52, 0.2)';
export const im_btn_selected_bg_green_press = 'rgba(65, 158, 52, 0.36)';
export const im_btn_selected_bg_red_hover = 'rgba(240, 91, 86, 0.28)';
export const im_btn_selected_bg_red_normal = 'rgba(240, 91, 86, 0.2)';
export const im_btn_selected_bg_red_press = 'rgba(240, 91, 86, 0.36)';
export const im_btn_selected_line_gray = 'rgba(255, 255, 255, 0.5)';
export const im_btn_selected_text_green = '#51ba43';
export const im_btn_selected_text_red = '#f05b56';
export const imtoken_AI_Profile_Avatar_bg = 'linear-gradient(to top, #292929 32%, #7157a8 66%, #b57cd9 100%)';
export const imtoken_AI_message_bg = '#151515';
export const imtoken_AI_onboarding_Avatar_bg = 'linear-gradient(to bottom, rgba(71, 82, 230, 0) 24%, rgba(71, 82, 230, 0.2) 75%, rgba(208, 95, 208, 0.5) 100%)';
export const imtoken_add_menu_icon_bg = '#373737';
export const imtoken_bg_Aurora_Horizon = '#0c0c0c';
export const imtoken_dialog_bg_top = 'linear-gradient(180deg, #16193a 0%, #292929 100%)';
export const imtoken_dialog_bg_top_cricle = 'linear-gradient(180deg, rgba(211, 229, 255, 0.06) 0%, rgba(41, 41, 41, 0.08) 100%)';
export const imtoken_feed_bg_body = '#757575';
export const imtoken_feed_bg_complete = '#1fa18f';
export const imtoken_feed_bg_delete = '#d14642';
export const imtoken_feed_bg_handle_later = '#f0b622';
export const imtoken_feed_bg_reaction = 'rgba(235, 235, 235, 0.1)';
export const imtoken_feed_bg_top = '#5f5f5f';
export const imtoken_feed_fill_active = 'rgba(76, 136, 255, 0.2)';
export const imtoken_feed_icon_pri_selected = '#ffffff';
export const imtoken_feed_item_active_btn_default = 'rgba(235, 235, 235, 0.15)';
export const imtoken_feed_item_active_btn_hover = 'rgba(235, 235, 235, 0.25)';
export const imtoken_feed_item_active_btn_press = 'rgba(235, 235, 235, 0.35)';
export const imtoken_feed_item_bg_press = 'rgba(235, 235, 235, 0.08)';
export const imtoken_feed_item_default_btn_default = '#373737';
export const imtoken_feed_item_hover_btn_default = '#404040';
export const imtoken_feed_item_hover_btn_hover = '#525252';
export const imtoken_feed_item_hover_btn_press = '#646464';
export const imtoken_feed_text_pri_selected = '#ffffff';
export const imtoken_feed_topitem_active_btn_default = 'rgba(235, 235, 235, 0.12)';
export const imtoken_feed_topitem_active_btn_hover = 'rgba(235, 235, 235, 0.2)';
export const imtoken_feed_topitem_active_btn_press = 'rgba(235, 235, 235, 0.3)';
export const imtoken_feed_topitem_bg_default = '#292929';
export const imtoken_feed_topitem_bg_hover = '#373737';
export const imtoken_feed_topitem_bg_press = '#3f3f3f';
export const imtoken_feed_topitem_default_btn_default = '#474747';
export const imtoken_feed_topitem_hover_btn_default = '#474747';
export const imtoken_feed_topitem_hover_btn_hover = '#575757';
export const imtoken_feed_topitem_hover_btn_press = '#676767';
export const imtoken_general_icon_encrypted_chat_normal = '#373737';
export const imtoken_message_bg_audio_blue = 'rgba(117, 164, 255, 0.15)';
export const imtoken_message_bg_audio_grey = 'rgba(235, 235, 235, 0.05)';
export const imtoken_message_bg_bubbles_blue = '#133063';
export const imtoken_message_bg_bubbles_grey = '#262626';
export const imtoken_message_bg_chat = '#151515';
export const imtoken_message_bg_chat_simple = '#1a1a1a';
export const imtoken_message_bg_location = '#63470f';
export const imtoken_message_bg_pin = '#473409';
export const imtoken_message_bg_reaction_blue = '#102954';
export const imtoken_message_bg_reaction_grey = '#1f1f1f';
export const imtoken_message_bg_reaction_grey_body = '#101010';
export const imtoken_message_bg_reaction_grey_simple = 'rgba(255, 255, 255, 0.05)';
export const imtoken_message_bg_selected = 'rgba(255, 255, 255, 0.05)';
export const imtoken_message_card_bg_body = '#373737';
export const imtoken_message_card_bg_body_embed = '#373737';
export const imtoken_message_card_border = 'rgba(235, 235, 235, 0)';
export const imtoken_message_icon_lan = '#f3871b';
export const imtoken_message_icon_pin = '#1fa18f';
export const imtoken_message_icon_pin_list = '#1ab7a1';
export const imtoken_message_icon_read = '#1fa18f';
export const imtoken_message_icon_unread = '#757575';
export const imtoken_message_icon_voice_blue = '#ffffff';
export const imtoken_message_icon_voice_blue_bg = '#4c88ff';
export const imtoken_message_icon_voice_grey = '#ffffff';
export const imtoken_message_icon_voice_grey_bg = '#757575';
export const imtoken_message_pin_list_popover_caret = '#35302c';
export const imtoken_message_pin_popover_bg = '#192a4c';
export const imtoken_message_pin_popover_line = '#3c3c3c';
export const imtoken_message_pin_text_bg_focus = 'rgba(26, 183, 161, 0.1)';
export const imtoken_message_pin_text_bg_hover = 'rgba(26, 183, 161, 0.1)';
export const imtoken_message_pin_text_bg_pressed = 'rgba(26, 183, 161, 0.2)';
export const imtoken_message_sidebar_group_bg = '#1a1a1a';
export const imtoken_message_sidebar_group_bg_simple = '#1e1e1e';
export const imtoken_message_text_bubbles_blue = '#75a4ff';
export const imtoken_message_text_bubbles_grey = '#cfcfcf';
export const imtoken_message_text_pin = '#1fa18f';
export const imtoken_message_voice_line_schedule_blue = '#8fb4ff';
export const imtoken_message_voice_line_schedule_grey = '#e0e0e0';
export const imtoken_message_voice_text_time_blue = '#8fb4ff';
export const imtoken_message_voice_text_time_gray = '#e0e0e0';
export const imtoken_profile_bg_hover = '#434343';
export const imtoken_profile_bg_normal = '#373737';
export const imtoken_profile_bg_press = '#434343';
export const imtoken_profile_icon_normal = '#e0e0e0';
export const imtoken_profile_text_normal = '#e8e8e8';
export const imtoken_red_envelope_button_text_yellow = '#ffeaa3';
export const imtoken_red_envelope_card_text_placeholder_yellow = 'rgba(255, 222, 117, 0.8)';
export const imtoken_red_envelope_card_text_title_yellow = '#ffde75';
export const imtoken_secret_chat_bg_topbar = '#383b42';
export const imtoken_secure_badge_bg_normal = '#434343';
export const imtoken_set_colour_Aurora_Air = 'linear-gradient(45deg, #302c44 0%, #2b3740 100%)';
export const imtoken_set_colour_Aurora_Horizon = '#050505';
export const imtoken_set_colour_Aurora_Lake = 'linear-gradient(45deg, #30415b 0%, #2f4463 100%)';
export const imtoken_set_colour_Aurora_Obsidian = 'linear-gradient(45deg, #333946 0%, #3b4252 100%)';
export const imtoken_set_colour_Aurora_Sky = '#2a333c';
export const imtoken_side_unread_bg_active = '#343d4f';
export const imtoken_side_unread_bg_default = '#262626';
export const imtoken_side_unread_bg_hover = '#363636';
export const imtoken_t_entrance_default = 'rgba(0, 0, 0, 0.4)';
export const imtoken_t_entrance_hover = 'rgba(0, 0, 0, 0.6)';
export const imtoken_text_tips_link_hover = '#4c88ff';
export const imtoken_text_tips_link_normal = '#75a4ff';
export const imtoken_text_tips_link_pressed = '#8fb4ff';
export const imtoken_topic_group_comment_file = '#373737';
