// @universe-design/react-utilities/lib-commonjs/descendant/use-descendant.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createDescendantContext = createDescendantContext;
exports.createDescendantContextOptional = createDescendantContextOptional;
var _react = _interopRequireWildcard(require("react"));
var _useIsomorphicLayoutEffect = require("../hooks/useIsomorphicLayoutEffect");
var _useForceUpdate = require("../hooks/useForceUpdate");
var _composeRefs = require("../hooks/composeRefs");
var _reactProvider = require("@universe-design/react-provider");
var _descendant = require("./descendant");
var _useConstant = require("../hooks/useConstant");
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/naming-convention */
/* eslint-disable @typescript-eslint/no-magic-numbers */

/**
 * @internal
 * React hook that initializes the DescendantsManager
 */
function createUseDescendants() {
  return function () {
    const descendants = (0, _useConstant.useConstant)(() => new _descendant.DescendantsManager());
    (0, _useIsomorphicLayoutEffect.useIsomorphicLayoutEffect)(() => () => descendants.destroy(), [descendants]);
    return descendants;
  };
}
function createUseDescendant(useDescendantsContext) {
  return function (options) {
    var _descendants$enabledI;
    const descendants = useDescendantsContext();
    const forceUpdate = (0, _useForceUpdate.useForceUpdate)();
    const [index, setIndex] = (0, _react.useState)(-1);
    const ref = (0, _react.useRef)(null);
    // 只在 unmount 时执行一次
    (0, _useIsomorphicLayoutEffect.useIsomorphicLayoutEffect)(() => () => {
      if (!ref.current || !descendants) {
        return;
      }
      descendants.unregister(ref.current);
    }, [descendants]);
    // eslint-disable-next-line react-hooks/exhaustive-deps
    (0, _useIsomorphicLayoutEffect.useIsomorphicLayoutEffect)(() => {
      if (!ref.current) {
        return;
      }
      const dataIndex = Number(ref.current.dataset.index);
      if (index !== dataIndex && !Number.isNaN(dataIndex)) {
        setIndex(dataIndex);
      }
    });
    const register = _react.default.useMemo(() => (0, _composeRefs.composeRefs)(descendants === null || descendants === void 0 ? void 0 : descendants.register({
      forceUpdate
    }, options), ref), [descendants, forceUpdate, options]);
    return {
      descendants,
      index,
      enabledIndex: (_descendants$enabledI = descendants === null || descendants === void 0 ? void 0 : descendants.enabledIndexOf(ref.current)) !== null && _descendants$enabledI !== void 0 ? _descendants$enabledI : -1,
      register
    };
  };
}
/**
 * Create a generic typed descendant context
 *
 * @public
 */
function createDescendantContext() {
  const Context = (0, _reactProvider.createContextHook)('DescendantsContext', {
    strict: true
  });
  return [
  // context provider
  Context.Provider,
  // call this when you need to read from context
  Context.useContext,
  // descendants state information, to be called and passed to `ContextProvider`
  createUseDescendants(),
  // descendant index information
  createUseDescendant(Context.useContext)];
}
/**
 * Create a generic typed descendant context
 *
 * @public
 */
function createDescendantContextOptional() {
  const Context = (0, _reactProvider.createContextHook)('DescendantContext(Optional)', {
    strict: false,
    defaultValue: null
  });
  return [
  // context provider
  Context.Provider,
  // call this when you need to read from context
  Context.useContext,
  // descendants state information, to be called and passed to `ContextProvider`
  createUseDescendants(),
  // descendant index information
  createUseDescendant(Context.useContext)];
}
