// @universe-design/styles/esm/mail/index.light.atom.js
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
export const mailtoken_attachment_card_bg_blue = 'linear-gradient(to right, #e0e6ff 0%, #f8f7fe 100%)';
export const mailtoken_attachment_card_bg_green = 'linear-gradient(to right, #def6d9 0%, #f4faea 100%)';
export const mailtoken_attachment_card_bg_grey = 'linear-gradient(to right, #eaeef3 0%, #f6f7fa 100%)';
export const mailtoken_attachment_card_bg_orange = 'linear-gradient(to right, #ffeade 0%, #fff7ec 100%)';
export const mailtoken_attachment_card_bg_red = 'linear-gradient(to right, #ffe9e8 0%, #fff7f2 100%)';
export const mailtoken_attachment_card_bg_violet = 'linear-gradient(to right, #ffe5ff 0%, #fff3fa 100%)';
export const mailtoken_attachment_card_bg_yellow = 'linear-gradient(to right, #ffeed0 0%, #fef7e7 100%)';
export const mailtoken_compose_default_bubble_bg_hover = '#dee0e3';
export const mailtoken_compose_default_bubble_bg_normal = '#eff0f1';
export const mailtoken_compose_default_bubble_bg_selected = '#e0e9ff';
export const mailtoken_compose_error_bubble_bg_hover = '#fdc6c4';
export const mailtoken_compose_error_bubble_bg_normal = '#fee3e2';
export const mailtoken_compose_error_bubble_bg_selected = '#e22e28';
export const mailtoken_compose_error_bubble_text_hover = '#a11c17';
export const mailtoken_compose_error_bubble_text_normal = '#c02a26';
export const mailtoken_compose_font_size_bg_selected = '#ffffff';
export const mailtoken_editor_tool_bg = '#ffffff';
export const mailtoken_follow_banner_bg = 'linear-gradient(92deg, #e4f3fa 0%, #e3edff 46.3%, #eceaff 100%)';
export const mailtoken_read_msglist_bg = '#f2f3f5';
