// @universe-design/react/lib/timeline/TimelineItem.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _TimeFilled = _interopRequireDefault(require("@universe-design/icons-react/TimeFilled"));
var _TimeOutlined = _interopRequireDefault(require("@universe-design/icons-react/TimeOutlined"));
var _CloseFilled = _interopRequireDefault(require("@universe-design/icons-react/CloseFilled"));
var _WarningFilled = _interopRequireDefault(require("@universe-design/icons-react/WarningFilled"));
var _YesFilled = _interopRequireDefault(require("@universe-design/icons-react/YesFilled"));
var _YesOutlined = _interopRequireDefault(require("@universe-design/icons-react/YesOutlined"));
var _hooks = require("../config-provider/hooks");
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _stylex2 = require("../stylex");
var _svgs = require("../_assets/svgs");
const useTimelineItemCssScope = () => (0, _hooks.useCssScope)('timeline-item');
const iconStatusMap = {
  success: /*#__PURE__*/_react.default.createElement(_YesFilled.default, null),
  warning: /*#__PURE__*/_react.default.createElement(_WarningFilled.default, null),
  error: /*#__PURE__*/_react.default.createElement(_CloseFilled.default, null),
  done: /*#__PURE__*/_react.default.createElement(_YesOutlined.default, null),
  current: /*#__PURE__*/_react.default.createElement(_TimeFilled.default, null),
  notarrived: /*#__PURE__*/_react.default.createElement(_TimeOutlined.default, null),
  loading: /*#__PURE__*/_react.default.createElement(_svgs.SpinLoadingSvg, null)
};
const TimelineItem = props => {
  const {
    position,
    labelPosition = 'relative',
    className,
    title,
    label,
    description,
    status,
    dotType = 'dot',
    dot = dotType === 'icon' && status && iconStatusMap[status],
    ...reset
  } = props;
  const itemCssScope = useTimelineItemCssScope();
  const labelEl = /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(itemCssScope('__label'), (0, _stylex.default)(_stylex2.cssTypographyBody2))
  }, label);
  const same = labelPosition === 'same';
  return /*#__PURE__*/_react.default.createElement("li", (0, _extends2.default)({}, reset, {
    className: (0, _clsx.default)(itemCssScope(), className, {
      [itemCssScope('--custom-dot')]: dot,
      [itemCssScope(`--position-${position}`)]: position,
      [itemCssScope(`--status-${status}`)]: status
    })
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: itemCssScope('__oppsite-content')
  }, !same && label && labelEl), /*#__PURE__*/_react.default.createElement("div", {
    className: itemCssScope('__seperator')
  }, dot ? /*#__PURE__*/_react.default.createElement("div", {
    className: itemCssScope('__dot--custom')
  }, dot) : /*#__PURE__*/_react.default.createElement("div", {
    className: itemCssScope('__dot')
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: itemCssScope('__tail')
  })), /*#__PURE__*/_react.default.createElement("div", {
    className: itemCssScope('__content')
  }, title && /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(itemCssScope('__title'), (0, _stylex.default)(_stylex2.cssTypographyBody0))
  }, title), same && label && labelEl, description && /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(itemCssScope('__des'), (0, _stylex.default)(_stylex2.cssTypographyBody2))
  }, description)));
};
if (process.env.NODE_ENV !== "production") {
  TimelineItem.displayName = 'Timeline.Item';
}
var _default = exports.default = TimelineItem;
