// @universe-design/react/lib/timeline/Timeline.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _hooks = require("../config-provider/hooks");
var _TimelineItem = _interopRequireDefault(require("./TimelineItem"));
var _util = require("../util");
var _loading = require("../loading");
require("./style");
const useTimelineCssScope = () => (0, _hooks.useCssScope)('timeline');
const useTimelineItemCssScope = () => (0, _hooks.useCssScope)('timeline-item');
const Timeline = props => {
  const {
    mode = 'left',
    reverse,
    hangingTail,
    children,
    pending,
    pendingDot,
    space = 'medium',
    style,
    className,
    labelPosition = 'relative',
    ...restProps
  } = props;
  const timelineItems = _react.default.Children.toArray(children);
  const validTimelineItems = timelineItems.filter(item => /*#__PURE__*/_react.default.isValidElement(item));
  const hasLabel = validTimelineItems.some(item => {
    var _item$props;
    return Boolean(item === null || item === void 0 ? void 0 : (_item$props = item.props) === null || _item$props === void 0 ? void 0 : _item$props.label);
  });
  const cssScope = useTimelineCssScope();
  const itemCssScope = useTimelineItemCssScope();
  const items = _react.default.Children.map(validTimelineItems, (item, index) => /*#__PURE__*/_react.default.cloneElement(item, {
    className: (0, _clsx.default)(mode === 'alternate' ? itemCssScope(`--mode-alt-${index % 2 === 0 ? 'left' : 'right'}`) : itemCssScope(`--mode-${mode}`), {
      [itemCssScope('-last')]: pending && !reverse && index === validTimelineItems.length - 1
    }, item.props.className),
    key: index,
    labelPosition
  }));
  if (pending) {
    const pendingNode = /*#__PURE__*/_react.default.createElement(_TimelineItem.default, {
      status: "current",
      className: (0, _clsx.default)(mode === 'alternate' ? itemCssScope('--mode-alt-left') : itemCssScope(`--mode-${mode}`), itemCssScope('--pending')),
      key: "pending",
      dot: pendingDot || /*#__PURE__*/_react.default.createElement(_loading.Loading, {
        className: itemCssScope('--loading')
      }),
      title: typeof pending === 'boolean' ? null : pending,
      labelPosition: labelPosition
    });
    items.push(pendingNode);
  }
  if (reverse) {
    items.reverse();
  }
  return /*#__PURE__*/_react.default.createElement("ul", (0, _extends2.default)({
    style: style
  }, restProps, {
    className: (0, _clsx.default)(cssScope(), {
      [cssScope('--no-label')]: !hasLabel,
      [cssScope('__hanging-tail')]: hangingTail,
      [cssScope('-pending')]: pending
    }, cssScope(`--space-${(0, _util.getDefaultSizeAbbr)(space, 'medium')}`), className)
  }), items);
};
Timeline.Item = _TimelineItem.default;
if (process.env.NODE_ENV !== "production") {
  Timeline.displayName = 'Timeline';
}
var _default = exports.default = Timeline;
