// @byted-service/bytedtrace/lib/serializer/byteWriteBuffer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ByteWriteBuffer = void 0;
const tslib_1 = require("tslib");
const utils = tslib_1.__importStar(require("./utils"));
const consts = tslib_1.__importStar(require("./consts"));
const logger_1 = require("../logger");
/**
 * 默认 1 m
 */
const defaultBufferSize = 1024 * 1024;
class ByteWriteBuffer {
    constructor(bigEndian, logger) {
        this.buffer = Buffer.alloc(defaultBufferSize);
        this.bigEndian = bigEndian;
        this.dataLength = 0;
        this.pool = {
            stringHeader: Buffer.alloc(2),
            textHeader: Buffer.alloc(5),
            intHeader: Buffer.alloc(5),
            longHeader: Buffer.alloc(9),
            uint64Header: Buffer.alloc(9),
            doubleHeader: Buffer.alloc(9),
            bytesUintHeader: Buffer.alloc(5),
            ipv4Header: Buffer.alloc(5),
            ipv6Header: Buffer.alloc(17),
            lengthBody: Buffer.alloc(4),
            traceLengthBody: Buffer.alloc(9),
            uuidHeader: Buffer.alloc(17),
            intLine: Buffer.alloc(4),
            uint64Line: Buffer.alloc(8),
        };
        this.pool.stringHeader[0] = consts.STRING_TYPE;
        this.pool.intHeader[0] = consts.INT_TYPE;
        this.pool.textHeader[0] = consts.TEXT_TYPE;
        this.pool.ipv4Header[0] = consts.IPV4_TYPE;
        this.pool.doubleHeader[0] = consts.DOUBLE_TYPE;
        this.pool.uint64Header[0] = consts.UINT64_TYPE;
        this.pool.longHeader[0] = consts.LONG_TYPE;
        this.pool.uuidHeader[0] = consts.UUID_TYPE;
        this.pool.ipv6Header[0] = consts.IPV6_TYPE;
        this.pool.bytesUintHeader[0] = consts.BYTES_TYPE;
        this.logger = logger || logger_1.defaultLogger;
    }
    /**
     * 检查buffer的有效长度与写入数据的长度
     * 如果buffer不够长，则进行扩容(长度暂时不受限制，所以一直会返回true)
     */
    isWritable(len) {
        const totalLength = this.len() + len;
        if (totalLength < this.buffer.length - this.dataLength) {
            return true;
        }
        else {
            /* istanbul ignore next */
            return this.expansion(len);
        }
    }
    /**
     * 扩展 buffer，每扩展一次增加 len + 1 M
     */
    expansion(len) {
        const totalLength = this.buffer.length + len + defaultBufferSize;
        const buffer = Buffer.alloc(totalLength);
        // 数据量变大时 copy 比 for forEach fill 都要快
        this.buffer.copy(buffer, 0);
        this.buffer = buffer;
        return true;
    }
    /**
     * 写入 buffer 返回写入 buffer 长度(单位：字节)
     * @param data
     * @returns
     */
    write(data) {
        if (this.isWritable(data.length)) {
            for (let i = 0; i < data.length; i++) {
                this.buffer[this.dataLength++] = data[i];
            }
            return data.length;
        }
        /* istanbul ignore next */
        return 0;
    }
    /**
     * 单独写入一个数据，没有key
     * 目前 writeValue 只用了 string 和 uint64
     * @param value
     * @param type
     * @returns
     */
    writeValue(value, type) {
        switch (type) {
            case 'string':
                return this.writeString(value);
            case 'uint64':
                return this.writeUint64(value);
            // case 'bytes':
            //     return this.writeBytes(value);
            // case 'int':
            //     return this.writeInt(value);
            // case 'uint':
            //     return this.writeUint(value);
            // case 'int8':
            //     return this.writeInt8(value);
            // case 'uint8':
            //     return this.writeUint8(value);
            // case 'int16':
            //     return this.writeInt16(value);
            // case 'uint16':
            //     return this.writeUint16(value);
            // case 'int32':
            //     return this.writeInt32(value);
            // case 'uint32':
            //     return this.writeUint32(value);
            // case 'int64':
            //     return this.writeInt64(value);
            // case 'float32' || 'float64':
            //     return this.writeFloat64(value);
            // case 'bool':
            //     return this.writeBool(value);
            // case 'serializer.Ipv6':
            //     return this.writeIpv6(value);
            // case 'serializer.Ipv4':
            //     return this.writeIpv4(value);
            // case 'serializer.Uuid':
            //     return this.writeUUid(value);
            default:
                /* istanbul ignore next */
                return 0;
        }
    }
    /**
     * 写入一个keyValue,key是一个字典
     * @param key
     * @param value
     * @param type
     * @returns
     */
    writeDicKeyValue(key, value, type) {
        switch (type) {
            case 'string':
                return this.writeDicStringValue(key, value);
            case 'bytes':
                return this.writeDicBytesValue(key, value);
            case 'int':
                return this.writeDicIntValue(key, value);
            case 'uint':
                return this.writeDicUintValue(key, value);
            case 'int8':
                return this.writeDicInt8Value(key, value);
            case 'uint8':
                return this.writeDicUint8Value(key, value);
            case 'int16':
                return this.writeDicInt16Value(key, value);
            case 'uint16':
                return this.writeDicUint16Value(key, value);
            case 'int32':
                return this.writeDicInt32Value(key, value);
            case 'uint32':
                return this.writeDicUint32Value(key, value);
            case 'int64':
                return this.writeDicInt64Value(key, value);
            case 'uint64':
                return this.writeDicUint64Value(key, value);
            case 'float32':
                return this.writeDicFloat32Value(key, value);
            case 'float64':
                return this.writeDicFloat64Value(key, value);
            case 'bool':
                return this.writeDicBooleanValue(key, value);
            case 'serializer.Ipv4':
                return this.writeDicIPV4Value(key, value);
            case 'serializer.Ipv6':
                return this.writeDicIPV6Value(key, value);
            case 'serializer.Uuid':
                return this.writeDicUUidValue(key, value);
            default:
                /* istanbul ignore next */
                throw new Error(`type: ${type} is unRecognize!`);
        }
    }
    /**
     * 写入 key 和对应 value
     * @param key
     * @param value
     * @param type
     * @returns
     */
    writeKeyValue(key, value, type) {
        switch (type) {
            case 'string':
                return this.writeStringValue(key, value);
            case 'bytes':
                return this.writeBytesValue(key, value);
            case 'int':
                return this.writeIntValue(key, value);
            case 'uint':
                return this.writeUintValue(key, value);
            case 'int8':
                return this.writeInt8Value(key, value);
            case 'uint8':
                return this.writeUint8Value(key, value);
            case 'int16':
                return this.writeInt16Value(key, value);
            case 'uint16':
                return this.writeUint16Value(key, value);
            case 'int32':
                return this.writeInt32Value(key, value);
            case 'uint32':
                return this.writeUint32Value(key, value);
            case 'int64':
                return this.writeInt64Value(key, value);
            case 'uint64':
                return this.writeUint64Value(key, value);
            case 'float32':
                return this.writeFloat32Value(key, value);
            case 'float64':
                return this.writeFloat64Value(key, value);
            case 'bool':
                return this.writeBoolValue(key, value);
            case 'serializer.Ipv4':
                return this.writeIpv4Value(key, value);
            case 'serializer.Ipv6':
                return this.writeIpv6Value(key, value);
            case 'serializer.Uuid':
                return this.writeUUidValue(key, value);
            default:
                /* istanbul ignore next */
                throw new Error(consts.UN_RECOGNIZE_TYPE_ERROR);
        }
    }
    /*istanbul ignore next */
    writeUint32(value) {
        return this.writeUint64(value + '');
    }
    /*istanbul ignore next */
    writeUint16(value) {
        return this.writeInt32(value);
    }
    /*istanbul ignore next */
    writeInt16(value) {
        return this.writeInt32(value);
    }
    /*istanbul ignore next */
    writeInt8(value) {
        return this.writeInt32(value);
    }
    /*istanbul ignore next */
    writeUint8(value) {
        return this.writeInt32(value);
    }
    /*istanbul ignore next */
    writeInt(value) {
        return this.writeInt64(value);
    }
    /*istanbul ignore next */
    writeUint(value) {
        return this.writeUint64(value + '');
    }
    /*istanbul ignore next */
    writeUUid(value) {
        utils.fillBufferWith0(this.pool.uuidHeader, 1);
        utils.appendToTarget(this.pool.uuidHeader, value);
        return this.write(this.pool.uuidHeader);
    }
    /**
     * @param data
     * @returns
     */
    /*istanbul ignore next */
    writeInt64(data) {
        const b = utils.int64ToBytes(data, this.bigEndian);
        utils.appendToTarget(this.pool.longHeader, b);
        return this.write(this.pool.longHeader);
    }
    writeUint64(value) {
        const bl = utils.uint64ToBytes(value, this.bigEndian);
        utils.fillBufferWith0(this.pool.uint64Header, 1);
        utils.appendToTarget(this.pool.uint64Header, bl);
        return this.write(this.pool.uint64Header);
    }
    /*istanbul ignore next */
    writeInt32(value) {
        const bl = utils.int32ToBytes(value, this.bigEndian);
        utils.appendToTarget(this.pool.intHeader, bl);
        return this.write(this.pool.intHeader);
    }
    /*istanbul ignore next */
    writeFloat64(value) {
        const bl = utils.float64ToBytes(value, this.bigEndian);
        utils.appendToTarget(this.pool.doubleHeader, bl);
        return this.write(this.pool.doubleHeader);
    }
    /*istanbul ignore next */
    writeIpv4(value) {
        const bl = utils.uint32ToBytes(value, this.bigEndian);
        utils.appendToTarget(this.pool.ipv4Header, bl);
        return this.write(this.pool.ipv4Header);
    }
    /*istanbul ignore next */
    writeIpv6(value) {
        utils.fillBufferWith0(this.pool.ipv6Header, 1);
        utils.appendToTarget(this.pool.ipv6Header, value);
        return this.write(this.pool.ipv6Header);
    }
    /**
     * 获取所有写入的数据
     * @returns
     */
    bytes() {
        const data = this.buffer.slice(0, this.dataLength);
        return data;
    }
    /**
     * 返回的数组为复制
     * @returns
     */
    bytesCopy() {
        return Buffer.from(this.bytes());
    }
    /**
     * 获取已写入数据的长度
     * @returns
     */
    len() {
        return this.dataLength;
    }
    /**
     * 数据清零
     */
    reset() {
        this.buffer.fill(0, 0, this.dataLength);
        this.dataLength = 0;
    }
    /**
     * 修改从post开始，4个字节表示的长度的值
     * @param pos
     * @param value
     */
    setPosValue(pos, value) {
        if (pos + 4 > this.buffer.length) {
            /* istanbul ignore next */
            throw new Error(consts.INDEX_OUT_OF_BOUND_ERROR);
        }
        const bytes = utils.intToBytes(value, this.bigEndian);
        logger_1.debugLogger.debug('setPosValue value:%d,byte:%o', value, bytes);
        for (let i = 0; i <= 3; i++) {
            this.buffer[pos + i] = bytes[i];
        }
    }
    writeDicInt8Value(key, value) {
        return this.writeDicInt32Value(key, value);
    }
    writeDicUint8Value(key, value) {
        return this.writeDicInt32Value(key, value);
    }
    writeDicInt16Value(key, value) {
        return this.writeDicInt32Value(key, value);
    }
    writeDicUint16Value(key, value) {
        return this.writeByte(key) + this.writeInt32(value);
    }
    writeDicInt32Value(key, value) {
        return this.writeByte(key) + this.writeInt32(value);
    }
    writeDicUint32Value(key, value) {
        return this.writeDicUint64Value(key, value + '');
    }
    writeDicInt64Value(key, value) {
        return this.writeByte(key) + this.writeInt64(value);
    }
    writeDicUint64Value(key, value) {
        return this.writeByte(key) + this.writeUint64(value);
    }
    writeDicFloat32Value(key, value) {
        return this.writeDicFloat64Value(key, value);
    }
    writeDicFloat64Value(key, value) {
        return this.writeByte(key) + this.writeDouble(value);
    }
    writeDicBooleanValue(key, value) {
        return this.writeByte(key) + this.writeBool(value);
    }
    writeDicIntValue(key, value) {
        return this.writeByte(key) + this.writeInt(value);
    }
    writeDicUintValue(key, value) {
        return this.writeDicUint64Value(key, value + '');
    }
    /**
     * 写入一个 double 到 buffer
     * @param value
     */
    writeDouble(value) {
        const bl = utils.float64ToBytes(value, this.bigEndian);
        utils.fillBufferWith0(this.pool.doubleHeader, 1);
        utils.appendToTarget(this.pool.doubleHeader, bl);
        return this.write(this.pool.doubleHeader);
    }
    /**
     * 写入一个字节到 buffer
     * @param value
     * @returns
     */
    writeByte(value) {
        if (this.isWritable(1)) {
            this.buffer.writeUInt8(value, this.dataLength);
            this.dataLength++;
            return 1;
        }
        /* istanbul ignore next */
        return 0;
    }
    writeBytes(value) {
        const bl = utils.intToBytes(value.length, this.bigEndian);
        utils.fillBufferWith0(this.pool.bytesUintHeader, 1);
        utils.appendToTarget(this.pool.bytesUintHeader, bl);
        this.write(this.pool.bytesUintHeader);
        return this.write(value) + 5;
    }
    writeDicStringValue(key, value) {
        return this.writeByte(key) + this.writeString(value);
    }
    writeDicBytesValue(key, value) {
        return this.writeByte(key) + this.writeBytes(value);
    }
    writeDicIPV4Value(key, value) {
        return this.writeByte(key) + this.writeIpv4(value);
    }
    writeDicIPV6Value(key, value) {
        return this.writeByte(key) + this.writeIpv6(value);
    }
    writeDicUUidValue(key, value) {
        return this.writeByte(key) + this.writeUUid(value);
    }
    writeStringValue(key, value) {
        return this.writeStringKey(key) + this.writeString(value);
    }
    writeBytesValue(key, value) {
        return this.writeStringKey(key) + this.writeBytes(value);
    }
    writeIntValue(key, value) {
        return this.writeStringKey(key) + this.writeInt(value);
    }
    writeUintValue(key, value) {
        return this.writeUint64Value(key, value + '');
    }
    writeInt8Value(key, value) {
        return this.writeInt32Value(key, value);
    }
    writeUint8Value(key, value) {
        return this.writeInt32Value(key, value);
    }
    writeInt16Value(key, value) {
        return this.writeInt32Value(key, value);
    }
    writeUint16Value(key, value) {
        return this.writeInt32Value(key, value);
    }
    writeInt32Value(key, value) {
        return this.writeStringKey(key) + this.writeInt32(value);
    }
    writeUint32Value(key, value) {
        return this.writeUint64Value(key, value + '');
    }
    writeInt64Value(key, value) {
        return this.writeStringKey(key) + this.writeInt64(value);
    }
    writeUint64Value(key, value) {
        return this.writeStringKey(key) + this.writeUint64(value);
    }
    writeFloat32Value(key, value) {
        return this.writeFloat64Value(key, value);
    }
    writeFloat64Value(key, value) {
        return this.writeStringKey(key) + this.writeDouble(value);
    }
    writeBoolValue(key, value) {
        return this.writeStringKey(key) + this.writeBool(value);
    }
    writeIpv4Value(key, value) {
        return this.writeStringKey(key) + this.writeIpv4(value);
    }
    writeIpv6Value(key, value) {
        return this.writeStringKey(key) + this.writeIpv6(value);
    }
    writeUUidValue(key, value) {
        return this.writeStringKey(key) + this.writeUUid(value);
    }
    // 写入一个string
    writeString(data) {
        const len = Buffer.byteLength(data);
        if (len <= consts.MAX_STRING_LENGTH) {
            return this.writeStringByte(data);
        }
        else {
            // 如果长度大于 255 字节，需要当做 text 写入
            return this.writeStringBytes(data);
        }
    }
    // 写入小于 256 字节的长度的字符
    writeStringByte(data) {
        utils.fillBufferWith0(this.pool.stringHeader, 1);
        this.pool.stringHeader[1] = data.length;
        const hl = this.write(this.pool.stringHeader);
        const wl = this.buffer.write(data, this.dataLength);
        this.dataLength += wl;
        this.writeByte(consts.STRING_SPLIT_BYTE);
        return hl + wl + 1;
    }
    // 写入大于 256 字节的长度的字符
    writeStringBytes(data) {
        const bl = utils.intToBytes(data.length, this.bigEndian);
        utils.fillBufferWith0(this.pool.textHeader, 1);
        utils.appendToTarget(this.pool.textHeader, bl);
        const hl = this.write(this.pool.textHeader);
        const wl = this.buffer.write(data, this.dataLength);
        this.dataLength += wl;
        this.writeByte(consts.STRING_SPLIT_BYTE);
        return hl + wl + 1;
    }
    // 写入一个 boolean
    /*istanbul ignore next */
    writeBool(data) {
        return this.write(data ? consts.TRUE_BOOL_DATA : consts.FALSE_BOOL_DATA);
    }
    // 写入String的key, key长度<256.
    writeStringKey(key) {
        const data = Buffer.from(key);
        // 长度大于 256 截断
        return this.writeStringByte(data.slice(0, consts.MAX_STRING_LENGTH).toString());
    }
}
exports.ByteWriteBuffer = ByteWriteBuffer;
