// @byted-service/bytedtrace/lib/serializer/dictionary.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Dictionary = void 0;
const tslib_1 = require("tslib");
const consts = tslib_1.__importStar(require("./consts"));
class Dictionary {
    /**
     * 数据清零
     */
    clear() {
        this.dic = new Map();
    }
    /**
     * 清除CommonKey以外的数据
     */
    reset() {
        this.commonDic = new Map();
        this.dic = new Map();
        this.decodeMap = new Map();
    }
    /**
     * 获取长度
     * @returns
     */
    len() {
        return this.commonDic.size + this.dic.size;
    }
    /**
     * 将字典展平
     * @returns
     */
    flatten() {
        const len = this.len();
        const result = [];
        const fill = (v, k) => {
            /* istanbul ignore if */
            if (v >= len) {
                throw new Error(consts.FLATTEN_DICTIONARY_ERROR);
            }
            result[v] = k;
        };
        for (const [k, v] of this.commonDic.entries()) {
            fill(v, k);
        }
        for (const [k, v] of this.dic.entries()) {
            fill(v, k);
        }
        return result;
    }
    /**
     * 添加一个公有的key
     * @param key
     * @returns
     */
    addCommonKey(key) {
        if (this.commonDic.has(key)) {
            return;
        }
        const len = this.len();
        if (len > consts.MAX_DICTIONARY_SIZE) {
            throw new Error(consts.OVER_DICTIONARY_SIZE_ERROR);
        }
        this.commonDic.set(key, len);
    }
    /**
     * 添加一个普通的key
     * @param key
     * @returns
     */
    addKey(key) {
        if (this.commonDic.has(key) || this.dic.has(key)) {
            return;
        }
        const len = this.len();
        /* istanbul ignore if */
        if (len > consts.MAX_DICTIONARY_SIZE) {
            throw new Error(consts.OVER_DICTIONARY_SIZE_ERROR);
        }
        this.dic.set(key, len);
    }
    /**
     * 获取一个key的编码
     * @param key
     * @returns
     */
    coding(key) {
        let id = this.commonDic.get(key);
        if (typeof id === 'number') {
            return id;
        }
        id = this.dic.get(key);
        if (typeof id === 'number') {
            return id;
        }
        /* istanbul ignore next */
        throw new Error(consts.DICTIONARY_HAS_NOT_KEY_ERROR);
    }
    /**
     * 根据平铺的keys构建字典
     * @param keys
     */
    build(keys) {
        if (keys.length > consts.MAX_DICTIONARY_SIZE) {
            throw new Error(consts.OVER_DICTIONARY_SIZE_ERROR);
        }
        keys.forEach((v, idx) => this.decodeMap.set(idx, v));
    }
    /**
     * 获取一个id对应的key
     * @param id
     * @returns
     */
    decode(id) {
        if (this.decodeMap.has(id)) {
            return this.decodeMap.get(id);
        }
        throw new Error(consts.OVER_DICTIONARY_SIZE_ERROR);
    }
}
exports.Dictionary = Dictionary;
