// @byted-service/bytedtrace/lib/serializer/flowSerializer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlowSerializer = void 0;
const tslib_1 = require("tslib");
/**
 * 上层 FlowSerializer 挂载一个 serializer 实例用作序列化
 */
const gConsts = tslib_1.__importStar(require("../consts"));
const consts = tslib_1.__importStar(require("./consts"));
const serializer_1 = require("./serializer");
class FlowSerializer {
    constructor(options, env) {
        this.options = options;
        const serializer = new serializer_1.Serializer(consts.FLOW_SERIALIZER_TYPE, {
            useDic: this.options.useDic,
            bigEndian: this.options.bigEndian,
            logger: options.logger,
        });
        serializer.setHeader({
            version: options.version,
            protoType: options.protoType,
            compression: options.compression,
            reserved: options.reserved,
            userId: options.userId,
            contentCompression: options.contentCompression,
        });
        serializer.setCommonHeaders([
            {
                key: gConsts.PSM_KEY,
                value: env.serviceName,
                type: 'string',
            },
            {
                key: gConsts.DC_KEY,
                value: env.dc,
                type: 'string',
            },
            {
                key: gConsts.CLUSTER_KEY,
                value: env.cluster,
                type: 'string',
            },
            {
                key: gConsts.DEPLOY_STAGE_KEY,
                value: env.stage,
                type: 'string',
            },
            {
                key: gConsts.IPV4_KEY,
                value: env.ipv4Uint32,
                type: 'uint32',
            },
            {
                key: gConsts.IPV6_KEY,
                value: env.ipv6Bytes,
                type: 'bytes',
            },
            {
                key: gConsts.LANGUAGE_KEY,
                value: env.language,
                type: 'string',
            },
            {
                key: gConsts.ENV_KEY,
                value: env.env,
                type: 'string',
            },
            {
                key: gConsts.POD_NAME_KEY,
                value: env.podName,
                type: 'string',
            },
            {
                key: gConsts.FRAMEWOKER,
                value: env.framework,
                type: 'string',
            },
        ]);
        this.serializer = serializer;
        this.dataSize = 0;
    }
    append(t) {
        const data = t.toBs(this.options.protoType);
        if (data) {
            if (data.length > t.tracer.options.breaker.config.maxTxnDataLen) {
                throw new Error(gConsts.TXN_LEN_OVER_FLOW);
            }
        }
        this.dataSize = this.dataSize + data.length;
        return this.feed(data, 'bytes');
    }
    size() {
        return this.dataSize;
    }
    feed(data, type) {
        return this.serializer.feed({
            logContent: {
                keyValues: [{ key: '_txn', value: data, type }],
            },
        });
    }
    getSerializedData() {
        const data = this.serializer.serialize();
        this.serializer.clear();
        this.dataSize = 0;
        return data;
    }
}
exports.FlowSerializer = FlowSerializer;
