// @byted-service/bytedtrace/lib/serializer/utils.js
"use strict";
/**
 * TODO 这里可能会非常频繁的创建 buffer 需要考虑池化
 * 调用 serialize 的方法时单线程操作，不存在异步，
 * 所以提前申请 buffer 用完 clear 应该也是可行的
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.appendToTarget = exports.bufferToInt = exports.fillBufferWith0 = exports.float64ToBytes = exports.uint32ToBytes = exports.int32ToBytes = exports.intToBytes = exports.uint64ToBytes = exports.int64ToBytes = exports.getBuffer4Byte = exports.getBuffer8Byte = void 0;
const int64_buffer_1 = require("int64-buffer");
/**
 * 创建 8 byte 的 buffer
 * @returns
 */
const getBuffer8Byte = () => {
    return Buffer.alloc(8);
};
exports.getBuffer8Byte = getBuffer8Byte;
/**
 * 创建 4 byte 的 buffer
 * @returns
 */
const getBuffer4Byte = () => {
    return Buffer.alloc(8);
};
exports.getBuffer4Byte = getBuffer4Byte;
const int64ToBytes = (long, bigEndian) => {
    long = String(long).valueOf();
    if (bigEndian) {
        return new int64_buffer_1.Int64BE(long).toBuffer(true);
    }
    else {
        return new int64_buffer_1.Int64LE(long).toBuffer(true);
    }
};
exports.int64ToBytes = int64ToBytes;
const uint64ToBytes = (long, bigEndian) => {
    long = String(long).valueOf();
    if (bigEndian) {
        return new int64_buffer_1.Uint64BE(long).toBuffer(true);
    }
    else {
        return new int64_buffer_1.Uint64LE(long).toBuffer(true);
    }
};
exports.uint64ToBytes = uint64ToBytes;
const intToBytes = (value, bigEndian) => {
    return (0, exports.uint32ToBytes)(value, bigEndian);
};
exports.intToBytes = intToBytes;
/**
 * int32 转成 uint32 后再转成 buffer
 * @param value
 * @param bigEndian
 * @returns
 */
const int32ToBytes = (value, bigEndian) => {
    return (0, exports.uint32ToBytes)(value, bigEndian);
};
exports.int32ToBytes = int32ToBytes;
/**
 * uint32 转成 buffer
 * @param value
 * @param bigEndian
 * @returns
 */
const uint32ToBytes = (value, bigEndian) => {
    const buffer = (0, exports.getBuffer4Byte)();
    if (bigEndian) {
        buffer.writeUInt32BE(value, 0);
    }
    else {
        buffer.writeUInt32LE(value, 0);
    }
    return buffer;
};
exports.uint32ToBytes = uint32ToBytes;
/**
 * float64 转成 buffer
 * @param value
 * @param bigEndian
 * @returns
 */
const float64ToBytes = (value, bigEndian) => {
    const buffer = (0, exports.getBuffer8Byte)();
    if (bigEndian) {
        buffer.writeDoubleBE(value, 0);
    }
    else {
        buffer.writeDoubleLE(value, 0);
    }
    return buffer;
};
exports.float64ToBytes = float64ToBytes;
/**
 * 从 start(包括start) 开始填充0
 * @param buffer
 * @param start
 * @returns
 */
const fillBufferWith0 = (buffer, start) => {
    return buffer.fill(0, start);
};
exports.fillBufferWith0 = fillBufferWith0;
/**
 * 转成十进制字符串
 * @param buff
 * @returns
 */
/* istanbul ignore next */
const bufferToInt = (buff) => {
    return [...buff].map((e) => Number(e).toString(10)).join(' ');
};
exports.bufferToInt = bufferToInt;
/**
 * 将 buffer append 到 target 中，从 start 这个点开始
 * @param target
 * @param buffer
 * @param start 起点
 */
const appendToTarget = (target, buffer, start = 1) => {
    for (let i = 0; i < buffer.length; i++) {
        target[i + start] = buffer[i];
    }
};
exports.appendToTarget = appendToTarget;
