// @universe-design/react/lib/cascader/CascaderOverlay.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CascaderOverlay = void 0;
var _react = _interopRequireWildcard(require("react"));
var _tree = require("../tree");
var _select = require("../select");
var _useCascade = _interopRequireDefault(require("./cascade/useCascade"));
var _cascadeUI = require("./cascade/cascadeUI");
var _CascadeContext = require("./cascade/CascadeContext");
var _SelectOverlayContainer = require("../select/SelectOverlayContainer");
var _clsx = _interopRequireDefault(require("clsx"));
var _common = require("./common");
const CascaderOverlay = props => {
  const cssScope = (0, _common.useCascaderMenuCssScope)();
  const {
    inputValue,
    valueSet,
    set,
    multiple
  } = (0, _react.useContext)(_select.SelectContext);
  const {
    doublyTreeNodeMap,
    doublyTreeNodeList,
    changeOnSelect,
    showCheckedStrategy,
    cascadeRef,
    ...restProps
  } = props;
  const onSelect = _react.default.useCallback((keys, _action) => {
    const action = {
      type: _action.selected ? _select.SelectActionType.Select : _select.SelectActionType.DeSelect,
      payload: _action.key
    };
    let closeOverlay = true;
    if (multiple) {
      closeOverlay = false;
    } else if (changeOnSelect && (0, _tree.isNonLeaf)(doublyTreeNodeMap[_action.key].__node__)) {
      closeOverlay = false;
    }
    set(keys, {
      closeOverlay,
      action
    });
  }, [changeOnSelect, doublyTreeNodeMap, multiple, set]);
  const onCheck = _react.default.useCallback((keys, _action) => {
    const action = {
      type: _action.checked ? _select.SelectActionType.Checked : _select.SelectActionType.UnChecked,
      payload: _action.key
    };
    set(keys, {
      action
    });
  }, [set]);
  const cascadeCtxValue = (0, _useCascade.default)({
    ...restProps,
    changeOnSelect,
    value: Array.from(valueSet),
    searchValue: inputValue,
    multiple,
    onSelect,
    onCheck,
    showCheckedStrategy
  }, {
    doublyTreeNodeMap,
    doublyTreeNodeList
  }, cascadeRef);
  const {
    searchFlattenTree
  } = cascadeCtxValue;
  const rootDoublyTreeNodeList = doublyTreeNodeList.filter(n => !n.parent);
  return /*#__PURE__*/_react.default.createElement(_CascadeContext.CascadeContextProvider, {
    value: cascadeCtxValue
  }, /*#__PURE__*/_react.default.createElement(_SelectOverlayContainer.SelectOverlayContainer, {
    className: (0, _clsx.default)(cssScope(), (inputValue || !rootDoublyTreeNodeList.length) && cssScope('-search') // 这个样式主要是为了设置 width 为 auto
    ),
    optionsLength: rootDoublyTreeNodeList.length,
    filteredOptionsLength: searchFlattenTree.length,
    menu: /*#__PURE__*/_react.default.createElement(_cascadeUI.CascadeUI, null)
  }));
};
exports.CascaderOverlay = CascaderOverlay;
