// @universe-design/react/lib/cascader/common.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getCascaderLabel = exports.defaultVirtual = exports.defaultFilterOption = void 0;
exports.ifCanBeSelected = ifCanBeSelected;
exports.useCascaderCssScope = useCascaderCssScope;
exports.useCascaderMenuCssScope = useCascaderMenuCssScope;
var _escapeRegExp = _interopRequireDefault(require("lodash/escapeRegExp"));
var _react = _interopRequireDefault(require("react"));
var _hooks = require("../config-provider/hooks");
var _shared = require("../shared");
var _tree = require("../tree");
var _util = require("../util");
function useCascaderCssScope() {
  return (0, _hooks.useCssScope)('cascader');
}
function useCascaderMenuCssScope() {
  return (0, _hooks.useCssScope)('cascader__menu');
}
const defaultFilterOption = function (inputValue, node) {
  const {
    __node__: {
      title,
      value,
      label
    }
  } = node;
  // eslint-disable-next-line no-nested-ternary
  const keyword = (0, _util.isString)(label) ? label : (0, _util.isString)(title) ? title : value;
  return !inputValue || new RegExp((0, _escapeRegExp.default)(inputValue), 'gi').test(keyword) || Boolean(node.parent && defaultFilterOption(inputValue, node.parent));
};
exports.defaultFilterOption = defaultFilterOption;
const getCascaderLabel = (node, options) => {
  const labelNodes = [];
  for (let c = node; c !== null; c = c.parent) {
    let labelContentNode = c.__node__.value;
    if (!(0, _util.isNonNullable)(options.optionLabelProp)) {
      const {
        title,
        label
      } = c.__node__;
      if ((0, _util.isNonNullable)(label)) {
        labelContentNode = label;
      } else if ((0, _util.isNonNullable)(title)) {
        labelContentNode = title;
      }
    } else {
      if (typeof options.optionLabelProp === 'string') {
        labelContentNode = c.__node__[options.optionLabelProp];
      } else {
        labelContentNode = options.optionLabelProp(c.__node__, {
          path: (0, _tree.getTreeNodeConfigPath)(c)
        });
      }
    }
    labelNodes.unshift(/*#__PURE__*/_react.default.createElement(_react.default.Fragment, {
      key: c.__node__.value
    }, /*#__PURE__*/_react.default.createElement(_shared.HighlightConsumer, {
      markRenderer: text => /*#__PURE__*/_react.default.createElement("span", {
        className: options.highlightMarkClassName
      }, text)
    }, labelContentNode)));
    if (!options.cascaded) {
      break;
    }
    if (c.parent) {
      labelNodes.unshift(' / ');
    }
  }
  return labelNodes;
};

// 一个节点是否能被选中
exports.getCascaderLabel = getCascaderLabel;
function ifCanBeSelected(_ref) {
  let {
    node,
    selectable,
    changeOnSelect,
    multiple
  } = _ref;
  /**
   * 1. 单选 select mode 父节点不可选
   * 2. 单选 select mode 父节点可选
   * 3. 多选 select mode 父节点不可选
   * 4. 多选 select mode 父节点可选（不存在）
   */
  return selectable && ((0, _tree.isLeaf)(node) || changeOnSelect) && !(multiple && (0, _tree.isNonLeaf)(node) && changeOnSelect);
}

// 默认开启虚拟滚动
const defaultVirtual = () => true;
exports.defaultVirtual = defaultVirtual;
