// @universe-design/react/lib/cascader/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Cascader = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _common = require("./common");
var _select = require("../select");
var _tree = require("../tree");
var _util = require("../util");
var _CascaderOverlay = require("./CascaderOverlay");
var _useCascaderData = require("./cascade/useCascaderData");
require("./style");
var _useCascade = require("./cascade/useCascade");
var _TreeCheckedContext = require("../tree/hooks/TreeCheckedContext");
var _configProvider = require("../config-provider");
/* eslint-disable react-hooks/rules-of-hooks */

const RefSelect = (0, _select.selectGenerator)({
  selectorCssScope: 'select',
  dropdownCssScope: 'cascader'
  // defaultOptionLabelProp: option => option.title || option.value,
});
const CascaderComp = /*#__PURE__*/_react.default.forwardRef(function (props, ref) {
  var _baseSelectProps$mode, _baseSelectProps$chan;
  const {
    options = [],
    filterOption,
    filterResultSorter,
    filterResultLimit,
    filterResultRenderer,
    checkable,
    listHeight,
    dropdownLevelWidth,
    expandTrigger,
    fullPathBackfillOption = false,
    optionLabelProp: optionLabelPropProp,
    loadedKeys,
    loadData,
    onLoad,
    showCheckedStrategy = _tree.ShowCheckedStrategy.SHOW_CHILD,
    fullHeight,
    virtual,
    ...baseSelectProps
  } = props;
  const mode = (_baseSelectProps$mode = baseSelectProps.mode) !== null && _baseSelectProps$mode !== void 0 ? _baseSelectProps$mode : 'single';
  const selectRef = (0, _react.useRef)(null);
  const cascadeRef = (0, _react.useRef)(null);
  const {
    doublyTreeNodeList,
    doublyTreeNodeMap,
    internalOptions
  } = (0, _useCascaderData.useCascaderData)({
    options
  });
  const treeCheckedContext = (0, _tree.useTreeCheck)({
    doublyTreeNodeMap,
    defaultCheckedKeys: (0, _util.genArray)(props.defaultValue),
    ...('value' in props ? {
      checkedKeys: (0, _util.genArray)(props.value)
    } : {}),
    checkStrategy: _useCascade.strategyMap[showCheckedStrategy],
    checkStrictly: false
  }, {
    skipped: !checkable
  });
  const {
    checkedSet
  } = treeCheckedContext;
  const checkedKeys = _react.default.useMemo(() => Object.keys(checkedSet).filter(key => checkedSet[key]), [checkedSet]);
  if (mode === 'multiple' && checkable) {
    if ('value' in props) {
      // 受控模式
      baseSelectProps.value = checkedKeys;
    } else {
      // 非受控模式
      baseSelectProps.defaultValue = checkedKeys;
    }
  }

  /**
   * Select Props
   */
  const optionLabelProp = (0, _react.useMemo)(() => _ref => {
    let {
      value
    } = _ref;
    return doublyTreeNodeMap[value] ? (0, _common.getCascaderLabel)(doublyTreeNodeMap[value], {
      highlightMarkClassName: '',
      cascaded: fullPathBackfillOption,
      optionLabelProp: optionLabelPropProp
    }) : value;
  }, [doublyTreeNodeMap, fullPathBackfillOption, optionLabelPropProp]);
  baseSelectProps.mode = mode;
  const changeOnSelect = (_baseSelectProps$chan = baseSelectProps.changeOnSelect) !== null && _baseSelectProps$chan !== void 0 ? _baseSelectProps$chan : mode !== 'single';
  baseSelectProps.changeOnSelect = changeOnSelect;
  process.env.NODE_ENV !== "production" ? (0, _util.warning)(!(mode === 'single' && checkable), (0, _util.devWarning)('Cascader', 'Single check mode does not support.')) : void 0;
  process.env.NODE_ENV !== "production" ? (0, _util.warning)(!(mode === 'multiple' && !checkable && changeOnSelect), (0, _util.devWarning)('Cascader', 'Multiple select mode changeOnSelect=true does not support.')) : void 0;
  process.env.NODE_ENV !== "production" ? (0, _util.warning)(!(mode === 'multiple' && checkable && !changeOnSelect), (0, _util.devWarning)('Cascader', 'Multiple check mode changeOnSelect=false does not support.')) : void 0;
  const overlayNode = /*#__PURE__*/_react.default.createElement(_TreeCheckedContext.TreeCheckedProvider, {
    value: treeCheckedContext
  }, /*#__PURE__*/_react.default.createElement(_CascaderOverlay.CascaderOverlay, {
    cascadeRef: cascadeRef,
    loadData,
    loadedKeys,
    onLoad,
    dropdownLevelWidth,
    listHeight,
    filterOption,
    filterResultSorter,
    filterResultLimit,
    filterResultRenderer,
    expandTrigger,
    changeOnSelect,
    checkable,
    doublyTreeNodeList,
    doublyTreeNodeMap,
    showCheckedStrategy,
    fullHeight,
    virtual
  }));
  _react.default.useImperativeHandle(ref, () => ({
    blur() {
      var _selectRef$current;
      (_selectRef$current = selectRef.current) === null || _selectRef$current === void 0 ? void 0 : _selectRef$current.blur();
    },
    focus() {
      var _selectRef$current2;
      (_selectRef$current2 = selectRef.current) === null || _selectRef$current2 === void 0 ? void 0 : _selectRef$current2.focus();
    },
    get input() {
      var _selectRef$current3;
      return (_selectRef$current3 = selectRef.current) === null || _selectRef$current3 === void 0 ? void 0 : _selectRef$current3.input;
    },
    scrollIntoView() {
      var _cascadeRef$current;
      for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }
      return (_cascadeRef$current = cascadeRef.current) === null || _cascadeRef$current === void 0 ? void 0 : _cascadeRef$current.scrollIntoView(...args);
    }
  }), [cascadeRef, selectRef]);
  let selectProps;
  if (baseSelectProps.mode === 'multiple') {
    selectProps = {
      ...baseSelectProps,
      mode: 'multiple',
      options: internalOptions,
      overlayNode,
      onChange: (value, meta) => {
        if (baseSelectProps.onChange) {
          const path = value.length ? value.reduce((a, b, index) => {
            const _path = (0, _tree.getTreeNodeConfigPath)(doublyTreeNodeMap[b]);
            a[index] = _path;
            return a;
          }, []) : [];
          baseSelectProps.onChange(value, {
            ...meta,
            path
          });
        }
      }
    };
  } else {
    selectProps = {
      ...baseSelectProps,
      mode: 'single',
      options: internalOptions,
      overlayNode,
      onChange: (value, meta) => {
        if (baseSelectProps.onChange) {
          const path = (0, _tree.getTreeNodeConfigPath)((0, _util.isNonNullable)(value) ? doublyTreeNodeMap[value] : null);
          baseSelectProps.onChange(value, {
            ...meta,
            path
          });
        }
      }
    };
  }
  return /*#__PURE__*/_react.default.createElement(RefSelect, (0, _extends2.default)({}, selectProps, {
    optionLabelProp: optionLabelProp,
    ref: selectRef
  }));
});
const Cascader = exports.Cascader = Object.assign(CascaderComp, {
  SHOW_PARENT: _tree.ShowCheckedStrategy.SHOW_PARENT,
  SHOW_CHILD: _tree.ShowCheckedStrategy.SHOW_CHILD,
  [_configProvider.COMPONENT_VERSION_ID]: "3.28.3"
});
if (process.env.NODE_ENV !== "production") {
  Cascader.displayName = 'Cascader';
}
