// @universe-design/react/lib/shared/components/AutoSizer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AutoSizer = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactUtilities = require("@universe-design/react-utilities");
var _hooks = require("../../hooks");
var _composeRefs = _interopRequireDefault(require("../../util/composeRefs"));
/**
 * Forked and edited from https://github.com/bvaughn/react-virtualized-auto-sizer/blob/master/src/index.js
 * Thanks for bvaughn's work!
 */

const AutoSizer = exports.AutoSizer = /*#__PURE__*/_react.default.forwardRef((props, ref) => {
  /** Props */
  const {
    defaultWidth = 0,
    defaultHeight = 0,
    disableWidth = false,
    disableHeight = false,
    className,
    style,
    onResize,
    children
  } = props;

  /** State */
  const [width, setWidth] = (0, _react.useState)(defaultWidth);
  const [height, setHeight] = (0, _react.useState)(defaultHeight);

  /** Refs */
  const autoSizerRef = (0, _react.useRef)(null);
  const parentNodeRef = (0, _react.useRef)();
  const handleResize = (0, _hooks.useEventCallback)(entries => {
    const parentNode = parentNodeRef.current;
    for (const entry of entries) {
      if (entry.target !== parentNode || !parentNode) {
        continue;
      }
      const contentBoxSize = Array.isArray(entry.contentBoxSize) ? entry.contentBoxSize[0] : entry.contentBoxSize; // firefox 不是数组，safari < 15.4 没有 contentBoxSize

      const size = (() => {
        if (contentBoxSize) {
          return {
            width: contentBoxSize.inlineSize,
            height: contentBoxSize.blockSize
          };
        }
        const _width = parentNode.offsetWidth;
        const _height = parentNode.offsetHeight;
        const cstyle = getComputedStyle(parentNode);
        // TODO 暂时只支持 px
        const paddingLeft = parseInt(cstyle.paddingLeft, 10) || 0;
        const paddingRight = parseInt(cstyle.paddingRight, 10) || 0;
        const paddingTop = parseInt(cstyle.paddingTop, 10) || 0;
        const paddingBottom = parseInt(cstyle.paddingBottom, 10) || 0;
        return {
          width: _width - paddingLeft - paddingRight,
          height: _height - paddingTop - paddingBottom
        };
      })();
      const widthChanged = !disableWidth && width !== size.width;
      const heightChanged = !disableHeight && height !== size.height;
      if (widthChanged) {
        setWidth(size.width);
      }
      if (heightChanged) {
        setHeight(size.height);
      }
      if (widthChanged || heightChanged) {
        onResize === null || onResize === void 0 ? void 0 : onResize(size);
      }
    }
  });
  (0, _hooks.useIsomorphicLayoutEffect)(() => {
    var _parentNode$ownerDocu, _parentNode$ownerDocu2;
    const autoSizer = autoSizerRef.current;
    const parentNode = autoSizer === null || autoSizer === void 0 ? void 0 : autoSizer.parentNode;
    const HTMLElement = parentNode === null || parentNode === void 0 ? void 0 : (_parentNode$ownerDocu = parentNode.ownerDocument) === null || _parentNode$ownerDocu === void 0 ? void 0 : (_parentNode$ownerDocu2 = _parentNode$ownerDocu.defaultView) === null || _parentNode$ownerDocu2 === void 0 ? void 0 : _parentNode$ownerDocu2.HTMLElement;
    if (!parentNode || !HTMLElement || !(parentNode instanceof HTMLElement)) {
      return;
    }
    parentNodeRef.current = parentNode;
    const ro = (0, _reactUtilities.createResizeObserver)();
    ro.subscribe(handleResize);
    ro.observer.observe(parentNode, {
      box: 'content-box'
    });
    return () => {
      ro.unsubscribe(handleResize);
      ro.observer.disconnect();
    };
  }, [handleResize]);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: className,
    style: {
      overflow: 'visible',
      ...style
    },
    ref: (0, _react.useMemo)(() => (0, _composeRefs.default)(ref, autoSizerRef), [ref])
  }, children({
    width,
    height
  }));
});
