// @universe-design/react/lib/shared/components/FirstLineAlign.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FirstLineAlign = exports.EmptyElement = void 0;
var _u1gmpn6f = _interopRequireDefault(require("./../../.stylex/u1gmpn6f"));
var _ug24l6d = _interopRequireDefault(require("./../../.stylex/ug24l6d"));
var _uidkx = _interopRequireDefault(require("./../../.stylex/uidkx17"));
var _uotlknz = _interopRequireDefault(require("./../../.stylex/uotlknz"));
var _u1k2iffh = _interopRequireDefault(require("./../../.stylex/u1k2iffh"));
var _ux645fm = _interopRequireDefault(require("./../../.stylex/ux645fm"));
var _u8wgnjj = _interopRequireDefault(require("./../../.stylex/u8wgnjj"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _clsx = _interopRequireDefault(require("clsx"));
var React = _interopRequireWildcard(require("react"));
var _react2 = require("@universe-design/stylex/react");
var _stylex2 = require("../../stylex");
var _forwardRef = require("../../util/forwardRef");
var _configProvider = require("../../config-provider");
//.u1guyajx{align-items:baseline;}
//.u1jhi6b8{flex-grow:0;}
//.ui64zpg{flex-shrink:0;}
//.u1rmlqtg{flex-basis:auto;}
//.u1j7mq28{width:0px;}
//.u1a3p1vp{overflow:hidden;}
//.u14t3ns0{display:inline-block;}
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
/**
 * 用于撑开 flex 父级
 */
const EmptyElement = exports.EmptyElement = (0, _forwardRef.forwardHTMLRef)((props, ref) => (0, _react2.jsx)("span", (0, _extends2.default)({
  ref: ref
}, (0, _react2.useMergeComponentProps)(props, {
  className: (0, _configProvider.useCssScope)('empty-inline-element')(''),
  sx: [{
    "props": {
      "a9b677": _u1k2iffh.default,
      "Bpg54ce": _ux645fm.default,
      "mc9l5x": _u8wgnjj.default
    }
  }]
})), '\xa0'));
if (process.env.NODE_ENV !== "production") {
  EmptyElement.displayName = 'EmptyElement';
}

/**
 * 用于首行对齐
 */
const FirstLineAlign = exports.FirstLineAlign = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    beforeProps,
    children,
    className,
    sx = [],
    ...restProps
  } = _ref;
  const before = (() => {
    if (!beforeProps) {
      return null;
    }
    const {
      className: beforeClassName,
      children: beforeChildren,
      sx: beforeSx = [],
      ...beforeRestProps
    } = beforeProps;
    return (0, _react2.jsx)("div", (0, _extends2.default)({
      className: (0, _clsx.default)(beforeClassName, (0, _stylex.default)(_stylex2.cssFlexBase, {
        "props": {
          "Bt984gj": _u1gmpn6f.default,
          "Bh6795r": _ug24l6d.default,
          "Bnnss6s": _uidkx.default,
          "xawz": _uotlknz.default
        }
      }, ...beforeSx))
    }, beforeRestProps), beforeChildren, (0, _react2.jsx)(EmptyElement, null));
  })();
  return (0, _react2.jsx)("div", (0, _extends2.default)({
    className: (0, _clsx.default)(className, (0, _stylex.default)(_stylex2.cssFlexBase, {
      "props": {
        "Bt984gj": _u1gmpn6f.default
      }
    }, ...sx)),
    ref: ref
  }, restProps), before, children, (0, _react2.jsx)(EmptyElement, null));
});
