// @universe-design/react/lib/shared/components/Image.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useImage_unstable = exports.Image_unstable = void 0;
var _u1g072z = _interopRequireDefault(require("./../../.stylex/u1g072z0"));
var _uqyag7z = _interopRequireDefault(require("./../../.stylex/uqyag7z"));
var _um7gkow = _interopRequireDefault(require("./../../.stylex/um7gkow"));
var _u145lqgf = _interopRequireDefault(require("./../../.stylex/u145lqgf"));
var _u1mrxcor = _interopRequireDefault(require("./../../.stylex/u1mrxcor"));
var _udg2fzc = _interopRequireDefault(require("./../../.stylex/udg2fzc"));
var _u13d3rh = _interopRequireDefault(require("./../../.stylex/u13d3rh8"));
var _u1p85sbs = _interopRequireDefault(require("./../../.stylex/u1p85sbs"));
var _u8wgnjj = _interopRequireDefault(require("./../../.stylex/u8wgnjj"));
var _react = require("@universe-design/stylex/react");
var React = _interopRequireWildcard(require("react"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _reactUtilities = require("@universe-design/react-utilities");
var _function = require("../../util/function");
var _internal = require("../../internal");
var _hooks = require("../../config-provider/hooks");
//.u13uwng7{object-position:left top;}
//.u1ps3kmd{object-fit:cover;}
//.u1kle4es{object-fit:contain;}
//.u1plgu50{object-fit:none;}
//.u14xojzb{object-position:center;}
//.u1l02sjl{height:100%;}
//.uly5x3f{width:100%;}
//.u1ewtqcl{box-sizing:border-box;}
//.u14t3ns0{display:inline-block;}
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const loading = 'loading';
const error = 'error';
const loaded = 'loaded';
const useImageStylexStyles = (0, _react.makeStylexStyles)(_ref => {
  let {
    prefix
  } = _ref;
  return {
    base: {
      "props": {
        "B7ck84d": _u1p85sbs.default,
        "mc9l5x": _u8wgnjj.default
      },
      "alias": prefix
    },
    status: {
      [loading]: {
        "props": {},
        "alias": `${prefix}--loading`
      },
      [error]: {
        "props": {},
        "alias": `${prefix}--error`
      },
      [loaded]: {
        "props": {},
        "alias": `${prefix}--loaded`
      }
    },
    // 图片适配尺寸策略
    fit: {
      center: {
        "props": {
          "st4lth": _u145lqgf.default,
          "Ermj5k": _u1mrxcor.default,
          "Bqenvij": _udg2fzc.default,
          "a9b677": _u13d3rh.default
        },
        "alias": `${prefix}--fit-center`
      },
      contain: {
        "props": {
          "st4lth": _um7gkow.default,
          "Ermj5k": _u1mrxcor.default,
          "Bqenvij": _udg2fzc.default,
          "a9b677": _u13d3rh.default
        },
        "alias": `${prefix}--fit-contain`
      },
      default: {
        "props": {}
      },
      cover: {
        "props": {
          "st4lth": _uqyag7z.default,
          "Ermj5k": _u1mrxcor.default,
          "Bqenvij": _udg2fzc.default,
          "a9b677": _u13d3rh.default
        },
        "alias": `${prefix}--fit-cover`
      },
      none: {
        "props": {
          "st4lth": _u145lqgf.default,
          "Ermj5k": _u1g072z.default,
          "Bqenvij": _udg2fzc.default,
          "a9b677": _u13d3rh.default
        },
        "alias": `${prefix}--fit-none`
      }
    }
  };
});
const useImage_unstable = (props, ref) => {
  const {
    fit = 'default',
    sx,
    ...htmlProps
  } = props;
  const styles = useImageStylexStyles({
    prefix: (0, _hooks.useCssScope)('image')()
  });
  const [status, setStatus] = React.useState(loading);
  React.useMemo(() => {
    props.src;
    setStatus(loading);
  }, [props.src]);
  const imageProps = {
    ...htmlProps,
    ref,
    sx: [styles.base, styles.fit[fit], styles.status[status], sx],
    onLoad: (0, _reactUtilities.useEventCallback)((0, _function.composeEventHandlers)(props.onLoad, () => {
      setStatus(loaded);
    })),
    onError: (0, _reactUtilities.useEventCallback)((0, _function.composeEventHandlers)(props.onError, () => {
      setStatus(error);
    }))
  };
  return {
    state: {
      status
    },
    slotsProps: {
      root: imageProps
    }
  };
};
exports.useImage_unstable = useImage_unstable;
const Image_unstable = exports.Image_unstable = (0, _internal.forwardHTMLRef)((props, ref) => {
  const {
    slotsProps
  } = useImage_unstable(props, ref);
  return (0, _react.jsx)("img", slotsProps.root);
});
