// @universe-design/react/es/date-picker/hooks/useFormatAndParser.js
/**
 * 将 input 的输入值转换为 date
 * 将 value 转换为 string 提供给 input
 */
import isFunction from 'lodash/isFunction';
import { useLocale } from '../../config-provider/hooks';
import moment from 'moment';
function getDefaultFormat(options) {
  const {
    mode,
    showTime = false,
    presetFormats
  } = options;
  switch (mode) {
    case 'year':
      return presetFormats.year;
    case 'month':
      return presetFormats.month;
    case 'date':
    default:
      return showTime ? `${presetFormats.date} HH:mm:ss` : presetFormats.date;
  }
}
export default _ref => {
  let {
    format: formatProp,
    parse,
    mode,
    showTime
  } = _ref;
  const {
    locale: localeName,
    components: {
      datePicker: locale
    }
  } = useLocale();
  const presetFormats = {
    year: 'YYYY',
    month: locale.longMonthFormat,
    date: locale.longDateFormat
  };
  const defaultFormat = getDefaultFormat({
    mode,
    showTime,
    presetFormats
  });
  const format = formatProp !== null && formatProp !== void 0 ? formatProp : defaultFormat;
  return {
    formatValue: function (value) {
      if (!value) {
        return '';
      }
      if (isFunction(format)) {
        return format(value);
      }
      return value.clone().locale(localeName).format(format);
    },
    parseValue: function (value) {
      if (!value) {
        return null;
      }
      if (isFunction(parse)) {
        return parse(value);
      }
      /**
       * 对于 parser 而言，如果 format 类型为 (value: moment) => string
       * 此时是无法处理的 （设计上的一个漏洞）
       * 此时回退到 defaultFormat
       */
      const formatForParser = typeof format === 'string' ? format : defaultFormat;
      return moment(value, formatForParser, localeName, true);
    },
    format
  };
};
