// @universe-design/react/src/upload/stories/actions.stories.tsx
import React, { useState } from 'react';
import FileImageColorful from '@universe-design/icons-react/FileImageColorful';
import FilePdfColorful from '@universe-design/icons-react/FilePdfColorful';
import PreviewOutlined from '@universe-design/icons-react/PreviewOutlined';
import Upload from '../';
import { Button, Dialog, Space, message } from '../..';
import type { UploadFile } from '../uploadFile/interface';
import type { Meta } from '@storybook/react';

export default {
  title: 'Components/Upload/临时示例',
  component: Upload,
} satisfies Meta<typeof Upload>;

type PreviewInfo = { url?: string; name?: string };

function ImageViewer({ url, name, onClose }: PreviewInfo & { onClose: () => void }) {
  if (!url) {
    return null;
  }
  return (
    <Dialog title={name} onOk={onClose} onCancel={onClose} size="small" visible>
      <img src={url} width={200} />
    </Dialog>
  );
}

export const Actions = () => {
  const defaultFileList: UploadFile[] = [
    {
      uid: '1',
      name: '大猫.png',
      url: 'https://voffline.byted.org/download/tos/schedule//inspirecloud-cn-bytedance-internal/baas/tth168/dd6712ef6874b8c8_1626956933376.png',
    },
    {
      uid: '8',
      name: '附件名称最多显示一行.pdf',
      url: 'https://voffline.byted.org/download/tos/schedule//inspirecrrloud-cn-bytedance-internal/baas/tth168/dd6712ef6874b8c8_1626956933376.png',
    },
  ];

  function iconRender(file: UploadFile) {
    if (file.name.endsWith('pdf')) {
      return <FilePdfColorful />;
    } else if (file.name.endsWith('png')) {
      return <FileImageColorful />;
    }
  }
  const [previewInfo, setPreviewInfo] = useState<PreviewInfo | null>(null);
  const [dialogVisible, setDialogVisible] = useState(false);

  function onPreview(file: UploadFile) {
    if (!file.url) {
      return;
    }
    setPreviewInfo({ url: file.url, name: file.name });
    setDialogVisible(true);
  }

  function onDownload(file: UploadFile) {
    if (file.name.endsWith('pdf')) {
      message.error('不允许下载PDF');
      return false;
    }
    message.success('下载逻辑交给你了，控制台查看参数');
    console.log('onDownload能拿到以下参数', file);
  }

  function onRemove(file: UploadFile) {
    // 不允许删除PDF文件
    if (file.name.endsWith('pdf')) {
      message.error('不允许删除PDF文件！');
      return false;
    }
    message.success('非PDF文件，可以删除~');

    console.log('onDownload能拿到以下参数', file);
  }
  return (
    <Space direction="vertical">
      {dialogVisible && <ImageViewer {...previewInfo} onClose={() => setDialogVisible(false)} />}
      <div>
        <p>Demo1 自定义预览</p>
        <Upload
          defaultFileList={defaultFileList}
          iconRender={iconRender}
          action="https://cloudapi.bytedance.net/faas/services/tth168/invoke/hello"
          onPreview={onPreview}
          listType="text"
        >
          <Button> 点击上传</Button>
        </Upload>
      </div>
      <div>
        <p>Demo1 自定义下载逻辑</p>
        <Upload
          defaultFileList={defaultFileList}
          iconRender={iconRender}
          action="https://cloudapi.bytedance.net/faas/services/tth168/invoke/hello"
          onDownload={onDownload}
          listType="text"
        >
          <Button> 点击上传</Button>
        </Upload>
        <p>Demo3 拦截删除动作，不允许删除PDF文件</p>
        <Upload
          defaultFileList={defaultFileList}
          iconRender={iconRender}
          action="https://cloudapi.bytedance.net/faas/services/tth168/invoke/hello"
          onRemove={onRemove}
          listType="text"
        >
          <Button> 点击上传</Button>
        </Upload>
      </div>
      <div>
        <p>隐藏下载按钮，更改预览图标</p>
        <Upload
          showUploadList={{
            showDownloadIcon: false,
            previewIcon: file => <div data-testid="preview-test">{file.name}</div>,
            retryIcon: () => <PreviewOutlined />,
            downloadIcon: () => <PreviewOutlined />,
            errorIcon: () => <PreviewOutlined />,
            removeIcon: () => <PreviewOutlined />,
          }}
          defaultFileList={defaultFileList}
          multiple
        >
          <Button type="primary">选择文件</Button>
        </Upload>
      </div>
    </Space>
  );
};
