// @universe-design/react/src/upload/stories/homeless.stories.tsx
import React, { useState } from 'react';
import type { UploadItemRender, UploadFile, UploadChangeParam, UploadListType } from '@universe-design/react';
import {
  Select,
  CardFile,
  ComposableUpload,
  PictureFile,
  Switch,
  TextFile,
  toast,
  UploadTrigger,
  useStyleSheetOnce,
  useUploadContext,
} from '@universe-design/react';
import { Space } from '@universe-design/react';
import { Button, Upload } from '@universe-design/react';
import type { Meta } from '@storybook/react';
import AddColorful from '@universe-design/icons-react/AddColorful';
import FileImageColorful from '@universe-design/icons-react/FileImageColorful';
import FilePdfColorful from '@universe-design/icons-react/FilePdfColorful';

export default {
  title: 'Components/Upload/临时示例',
  component: Upload,
} satisfies Meta;

const itemRender: UploadItemRender = (originNode, file, currFileList, actions) => {
  const { name, status, uid, url } = file;
  const index = currFileList.indexOf(file);
  return (
    <div className="custom-item-render">
      <span>{`uid:${uid} name: ${name} status: ${status} url: ${url}  ${index + 1}/${currFileList.length}`}</span>
      <span onClick={actions.remove} className="custom-item-render-action-remove" data-testid="remove">
        remove
      </span>
      <span onClick={actions.download} className="custom-item-render-action-download" data-testid="download">
        download
      </span>
      <span onClick={actions.preview} className="custom-item-render-action-preview" data-testid="preview">
        preview
      </span>
    </div>
  );
};

export const UploadItemRenderDemo = function () {
  const defaultFileList: UploadFile[] = [
    {
      uid: '1',
      name: '大猫.png',
      url: 'https://voffline.byted.org/download/tos/schedule//inspirecloud-cn-bytedance-internal/baas/tth168/dd6712ef6874b8c8_1626956933376.png',
    },
    {
      uid: '8',
      name: '附件名称最多显示一行.pdf',
      url: 'https://voffline.byted.org/download/tos/schedule//inspirecrrloud-cn-bytedance-internal/baas/tth168/dd6712ef6874b8c8_1626956933376.png',
    },
  ];

  return (
    <Upload
      defaultFileList={defaultFileList}
      action="https://cloudapi.bytedance.net/faas/services/tth168/invoke/hello"
      itemRender={itemRender}
      onRemove={() => console.log('removed')}
    >
      <Button>上传</Button>
    </Upload>
  );
};

export function PicturesWall() {
  const fileList1: UploadFile[] = [
    {
      uid: '-2',
      name: 'image.png',
      status: 'success',
      url: 'https://tosv.byted.org/obj/inspirecloud-cn-bytedance-internal/baas/tth168/9eb3579cedd2d869_1630487982330.png',
    },
    {
      uid: '-32',
      name: 'image.png',
      status: 'waiting',
      url: 'https://tosv.byted.org/obj/inspirecloud-cn-bytedance-internal/baas/tth168/9eb3579cedd2d869_1630487982330.png',
    },
    {
      uid: '-xxx',
      percent: 50,
      name: 'image.png',
      status: 'uploading',
      url: 'https://tosv.byted.org/obj/inspirecloud-cn-bytedance-internal/baas/tth168/9eb3579cedd2d869_1630487982330.png',
    },
    {
      uid: '-5',
      name: 'image.png',
      status: 'error',
    },
    {
      uid: '-23',
      name: 'image.png',
      status: 'success',
      url: 'https://tosv.byted.org/obj/inspirecloud-cn-bytedance-internal/baas/tth168/9eb3579cedd2d869_1630487982330.png',
    },
    {
      uid: '-332',
      name: 'image.png',
      status: 'waiting',
      url: 'https://tosv.byted.org/obj/inspirecloud-cn-bytedance-internal/baas/tth168/9eb3579cedd2d869_1630487982330.png',
    },
    {
      uid: '-xxx4',
      percent: 50,
      name: 'image.png',
      status: 'uploading',
      url: 'https://tosv.byted.org/obj/inspirecloud-cn-bytedance-internal/baas/tth168/9eb3579cedd2d869_1630487982330.png',
    },
    {
      uid: '-54',
      name: 'image.png',
      status: 'error',
    },
  ];
  const fileList2: UploadFile[] = [
    {
      uid: '-2',
      name: 'image.png',
      status: 'success',
      url: 'https://tosv.byted.org/obj/inspirecloud-cn-bytedance-internal/baas/tth168/9eb3579cedd2d869_1630487982330.png',
    },

    {
      uid: '-xxx',
      percent: 50,
      name: 'image.png',
      status: 'success',
      url: 'https://tosv.byted.org/obj/inspirecloud-cn-bytedance-internal/baas/tth168/9eb3579cedd2d869_1630487982330.png',
    },
  ];
  const [fileList, setFileList] = useState<UploadFile[]>(fileList1);
  function onChange(info: UploadChangeParam) {
    const newFileList = info.fileList.map(file => {
      if (file.response) {
        file.url = file.response;
      }
      return file;
    });
    setFileList(newFileList);
  }
  return (
    <Space direction="vertical" style={{ padding: '30px' }}>
      <Upload
        action="https://cloudapi.bytedance.net/faas/services/tth168/invoke/hello"
        listType="picture"
        fileList={fileList}
        onChange={onChange}
        className="my-upload"
        showUploadList={{
          showPreviewIcon: true,
          showDownloadIcon: true,
        }}
        tips={
          <p
            style={{
              lineHeight: '22px',
              margin: 0,
              fontSize: '14px',
              color: '#8F959E',
            }}
          >
            请选择文件上传
          </p>
        }
        maxDisplayedPictureCount={3}
        multiple
      />
      <div>
        <p>禁用态：</p>
        <Upload
          action="https://cloudapi.bytedance.net/faas/services/tth168/invoke/hello"
          listType="picture"
          fileList={fileList2}
          onChange={onChange}
          multiple
          disabled
          maxDisplayedPictureCount={1}
        />
      </div>
    </Space>
  );
}

export function Dragger() {
  const [fileList, setFileList] = useState<UploadFile[]>([]);
  const props = {
    fileList,
    accept: 'image/*',
    action: 'https://cloudapi.bytedance.net/faas/services/tth168/invoke/hello',
    onChange(info: { file: UploadFile; fileList: UploadFile[] }) {
      setFileList(() =>
        (info.fileList || []).map(item => ({
          ...item,
          url: item.status === 'success' ? item.response || item.url : '',
        })),
      );
    },
  };

  return (
    <Space style={{ width: '536px', padding: '30px' }} direction="vertical">
      <div>
        <p>拖拽上传 DEMO 01</p>
        <Upload.Dragger
          {...props}
          dragAreaConfig={{
            dragTitle: (
              <p style={{ margin: 0 }}>
                拖拽文件至此，或
                <Button type="text" color="primary">
                  选择文件上传
                </Button>
              </p>
            ),
            dragDescription: '请选择图片文件, 支持格式：JPG、PNG、JPEG',
          }}
        />
      </div>
      <div>
        <p>拖拽上传 DEMO 02</p>

        <Upload.Dragger {...props}>
          <div className="ud__upload-drag-area">
            <div className="ud__upload-drop-area-title">拖拽文件至此，或点击上传</div>
          </div>
        </Upload.Dragger>
      </div>
      <div>
        <p>拖拽上传 DEMO 03</p>

        <Upload.Dragger
          {...props}
          listType="simple-text"
          dragAreaConfig={{
            dragTitle: '点这里可以上传',
            dragDescription: '请选择图片文件',
          }}
        />
      </div>
      <div>
        <p>拖拽上传 DEMO 04</p>

        <Upload.Dragger
          {...props}
          listType="card"
          dragAreaConfig={{
            dragTitle: (
              <p style={{ margin: 0 }}>
                拖拽文件至此，或
                <Button type="text" color="primary">
                  选择文件上传
                </Button>
              </p>
            ),
            dragIcon: <AddColorful />,
            dragDescription: '请选择图片文件, 支持格式：JPG、PNG、JPEG',
          }}
        />
      </div>
      <div>
        <p>拖拽上传: 禁用态</p>

        <Upload.Dragger
          {...props}
          listType="card"
          disabled
          dragAreaConfig={{
            dragTitle: (
              <p style={{ margin: 0 }}>
                拖拽文件至此，或
                <Button type="text" color="primary" disabled>
                  选择文件上传
                </Button>
              </p>
            ),
            dragIcon: <AddColorful />,
            dragDescription: '请选择图片文件, 支持格式：JPG、PNG、JPEG',
          }}
        />
      </div>
    </Space>
  );
}

const pictureFiles: UploadFile[] = [
  {
    uid: '-2',
    name: 'image.png',
    status: 'success',
    url: 'https://tosv.byted.org/obj/inspirecloud-cn-bytedance-internal/baas/tth168/9eb3579cedd2d869_1630487982330.png',
  },
  {
    uid: '-32',
    name: 'image.png',
    status: 'waiting',
    url: 'https://tosv.byted.org/obj/inspirecloud-cn-bytedance-internal/baas/tth168/9eb3579cedd2d869_1630487982330.png',
  },
  {
    uid: '-xxx',
    percent: 50,
    name: 'image.png',
    status: 'uploading',
    url: 'https://tosv.byted.org/obj/inspirecloud-cn-bytedance-internal/baas/tth168/9eb3579cedd2d869_1630487982330.png',
  },
  {
    uid: '-5',
    name: 'image.png',
    status: 'error',
  },
  {
    uid: '-23',
    name: 'image.png',
    status: 'success',
    url: 'https://tosv.byted.org/obj/inspirecloud-cn-bytedance-internal/baas/tth168/9eb3579cedd2d869_1630487982330.png',
  },
  {
    uid: '-332',
    name: 'image.png',
    status: 'waiting',
    url: 'https://tosv.byted.org/obj/inspirecloud-cn-bytedance-internal/baas/tth168/9eb3579cedd2d869_1630487982330.png',
  },
  {
    uid: '-xxx4',
    percent: 50,
    name: 'image.png',
    status: 'uploading',
    url: 'https://tosv.byted.org/obj/inspirecloud-cn-bytedance-internal/baas/tth168/9eb3579cedd2d869_1630487982330.png',
  },
  {
    uid: '-54',
    name: 'image.png',
    status: 'error',
  },
];

const defaultFileList: UploadFile[] = [
  {
    uid: '-1',
    name: 'xxx.png',
    status: 'success',
    url: 'https://tosv.byted.org/obj/inspirecloud-cn-bytedance-internal/baas/tth168/9eb3579cedd2d869_1630487982330.png',
    thumbUrl:
      'https://voffline.byted.org/download/tos/schedule//inspirecloud-cn-bytedance-internal/baas/tth168/dd6712ef6874b8c8_1626956933376.png',
  },
  {
    uid: '1',
    name: '大猫.png',
    status: 'success',
    url: 'https://voffline.byted.org/download/tos/schedule//inspirecloud-cn-bytedance-internal/baas/tth168/dd6712ef6874b8c8_1626956933376.png',
  },
  {
    uid: '3',
    name: '小猫咪.jpg',
    status: 'waiting',
    url: 'https://voffline.byted.org/download/tos/schedule//inspirecloud-cn-bytedance-internal/baas/tth168/67cb87f81e46730b_1626956991743.png',
    size: 434223,
  },
  {
    uid: '5',
    name: '附件名称最多显示一行附件名称最多显示一行附件名称最多显示一行.doc',
    percent: 80,
    status: 'uploading',
    size: 434223,
    loaded: 334223,
  },
  {
    uid: '83',
    name: '附件名称最多显示一行.pdf',
    status: 'error',
    response: 'Server Error 500', // custom error message to show
    failedReason:
      '上传失败，我感觉文件可能太大了,上传失败，我感觉文件可能太大了,上传失败，我感觉文件可能太大了.上传失败，我感觉文件可能太大了,上传失败，我感觉文件可能太大了,上传失败，我感觉文件可能太大了',
    size: 434223,
  },
  {
    uid: '8',
    name: '附件名称最多显示一行.pdf',
    status: 'error',
    response: 'Server Error 500', // custom error message to show
    size: 434223,
  },
];
export const FileListTypes = () => {
  const [list, setFileList] = useState<UploadFile[] | undefined>();

  function onChange({ fileList }: { fileList: UploadFile[] }) {
    const newFileList = fileList.map(file => {
      if (file.response) {
        file.url = file.response;
      }
      return file;
    });
    setFileList(newFileList);
  }

  function iconRender(file: UploadFile) {
    if (file.name.endsWith('pdf')) {
      return <FilePdfColorful />;
    } else if (file.name.endsWith('png')) {
      return <FileImageColorful />;
    }
  }
  return (
    <div style={{ width: '650px', padding: '30px' }}>
      {[null, 'text', 'simple-text', 'card'].map(type => (
        <div key={type}>
          <h1>ListType: {type || 'default'}</h1>
          <Upload
            defaultFileList={defaultFileList}
            iconRender={iconRender}
            action="https://cloudapi.bytedance.net/faas/services/tth168/invoke/hello"
            fileList={list}
            tips={
              <p
                style={{
                  lineHeight: '22px',
                  margin: 0,
                  fontSize: '14px',
                  color: '#8F959E',
                }}
              >
                请选择文件上传
              </p>
            }
            showUploadList={{
              showPreviewIcon: true,
              showDownloadIcon: true,
            }}
            onChange={onChange}
            listType={type as UploadListType}
          >
            <Button> 点击上传</Button>
          </Upload>
        </div>
      ))}
    </div>
  );
};

export const uploadProgress = () => {
  const fileList: UploadFile[] = [
    {
      uid: '5',
      name: '附件名称最多显示一行附件名称最多显示一行附件名称最多显示一行.doc',
      percent: 80,
      status: 'uploading',
      size: 434223,
      loaded: 334223,
    },
  ];

  return (
    <Upload
      defaultFileList={fileList}
      action="https://cloudapi.bytedance.net/faas/services/tth168/invoke/hello"
      progress={{
        strokeColor: 'red',
        strokeWidth: 2,
      }}
    />
  );
};

export const Files = () => {
  function iconRender(file: UploadFile) {
    if (file.name.endsWith('pdf')) {
      return <FilePdfColorful />;
    } else if (file.name.endsWith('png')) {
      return <FileImageColorful />;
    }
  }
  return (
    <>
      <p>Default Text FileList</p>
      <Space direction="vertical" size="small">
        {defaultFileList.map(file => (
          <TextFile
            file={file}
            disabled={false}
            onDownload={() => toast.success('download')}
            onPreview={() => toast.success('preview')}
            onRemove={() => toast.success('remove')}
            onRetry={() => {
              toast.success('retry');
            }}
            key={file.uid}
            iconRender={iconRender}
          />
        ))}
      </Space>
      <p>Simple Text FileList</p>
      <Space direction="vertical" size="small">
        {defaultFileList.map(file => (
          <TextFile
            file={file}
            disabled={false}
            onDownload={() => toast.success('download')}
            onPreview={() => toast.success('preview')}
            onRemove={() => toast.success('remove')}
            onRetry={() => {
              toast.success('retry');
            }}
            key={file.uid}
            iconRender={iconRender}
            isSimple
            progress={{
              strokeColor: 'red',
            }}
          />
        ))}
      </Space>
      <p>Simple Text FileList</p>
      <Space size="small">
        {defaultFileList.map(file => (
          <CardFile
            file={file}
            disabled={false}
            onDownload={() => toast.success('download')}
            onPreview={() => toast.success('preview')}
            onRemove={() => toast.success('remove')}
            onRetry={() => {
              toast.success('retry');
            }}
            key={file.uid}
            iconRender={iconRender}
          />
        ))}
      </Space>{' '}
      <p>Picture FileList</p>
      <Space size="small">
        {pictureFiles.map(file => (
          <PictureFile
            file={file}
            disabled={false}
            onDownload={() => toast.success('download', 0)}
            onPreview={() => toast.success('preview')}
            onRemove={() => toast.success('remove')}
            onRetry={() => {
              toast.success('retry');
            }}
            key={file.uid}
            iconRender={iconRender}
          />
        ))}
      </Space>
    </>
  );
};

export const TwoColumns = () => {
  const CustomFileList = () => {
    useStyleSheetOnce(`
      .file-container {
        display: flex;
        justify-content: space-between;
        flex-wrap: wrap;
        margin-top: 8px;
        max-width: 600px;
      }
      .file-item {
        width: 48%;
        margin-bottom: 8px;
      }
      `);
    const { fileList } = useUploadContext();

    return (
      <div className="file-container">
        {fileList.map(file => (
          <TextFile className="file-item" file={file} key={file.uid} isSimple />
        ))}
      </div>
    );
  };

  return (
    <ComposableUpload
      action="https://cloudapi.bytedance.net/faas/services/tth168/invoke/hello"
      defaultFileList={defaultFileList}
    >
      <UploadTrigger>
        <Button type="primary"> 上传 </Button>
      </UploadTrigger>
      <CustomFileList />
    </ComposableUpload>
  );
};

export function DifferentTypes() {
  const [fileList, setFileList] = useState<UploadFile[]>(defaultFileList);
  const [disabled, setDisabled] = useState(false);
  const [listType, setListType] = useState<UploadListType | undefined>('picture');
  function onChange(info: UploadChangeParam) {
    setFileList(() =>
      (info.fileList || []).map(item => ({
        ...item,
        url: item.url || (item.status === 'success' ? item.response : null),
      })),
    );
  }

  return (
    <>
      <Select<UploadListType>
        value={listType}
        options={[{ value: 'text' }, { value: 'card' }, { value: 'picture' }, { value: 'simple-text' }]}
        onChange={type => setListType(type)}
      />
      <p>
        Disabled: <Switch checked={disabled} onChange={v => setDisabled(v)} />
      </p>
      <Upload
        fileList={fileList}
        action="https://cloudapi.bytedance.net/faas/services/tth168/invoke/hello"
        onChange={onChange}
        listType={listType}
        disabled={disabled}
        maxDisplayedPictureCount={3}
      >
        <Button type="primary">选择文件</Button>
      </Upload>
    </>
  );
}

export function ItemRender() {
  const [fileList, setFileList] = useState([{ name: '123' }, { name: '123' }]);
  function onChange(info: UploadChangeParam) {
    setFileList(() =>
      (info.fileList || []).map(item => ({
        ...item,
        url: item.url || (item.status === 'success' ? item.response : null),
      })),
    );
  }
  return (
    <Upload
      listType="picture"
      action="https://cloudapi.bytedance.net/faas/services/tth168/invoke/hello"
      fileList={fileList}
      onChange={onChange}
      itemRender={(originNode, file) => <div style={{ border: '1px solid green' }}>{originNode}</div>}
    >
      <Button type="primary">选择文件</Button>
    </Upload>
  );
}
export function Icons() {
  const extNameFromLarkComponents = {
    doc: 'DOC',
    dot: 'DOC',
    docx: 'DOC',
    dotx: 'DOC',
    xls: 'XLS',
    xlt: 'XLS',
    xlsx: 'XLS',
    xltx: 'XLS',
    csv: 'XLS',
    xlsm: 'XLS',
    ppt: 'PPT',
    pptx: 'PPT',
    ppsx: 'PPT',
    ppam: 'PPT',
    potx: 'PPT',
    zip: 'ZIP',
    rar: 'ZIP',
    tar: 'ZIP',
    '7z': 'ZIP',
    pdf: 'PDF',
    jpg: 'IMG',
    jpeg: 'IMG',
    png: 'IMG',
    bmp: 'IMG',
    tiff: 'IMG',
    svg: 'IMG',
    rmvb: 'VIDEO',
    avi: 'VIDEO',
    mp4: 'VIDEO',
    qt: 'VIDEO',
    mov: 'VIDEO',
    wmv: 'VIDEO',
    mpeg: 'VIDEO',
    m4v: 'VIDEO',
    rm: 'VIDEO',
    flv: 'VIDEO',
    mkv: 'VIDEO',
    gif: 'IMG',
    ai: 'AI',
    psd: 'PSD',
    sketch: 'SKETCH',
    txt: 'TXT',
    apk: 'APK',
    mpga: 'AUDIO', // .mp3
    m4a: 'AUDIO',
    wav: 'AUDIO',
    aac: 'AUDIO',
    au: 'AUDIO',
    flac: 'AUDIO',
    oga: 'AUDIO', // .ogg
    wma: 'AUDIO',
    mid: 'AUDIO',
    aep: 'AEP',
  };

  const fileList = Object.keys(extNameFromLarkComponents).map(ext => ({
    name: `text.${ext}`,
  }));

  return <Upload fileList={fileList} listType="simple-text" />;
}
