// @universe-design/react/src/upload/stories/screenshot.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Basic from '../demo/0.basic';
import C1Filelist from '../demo/1.filelist';
import C10Form from '../demo/10.form';
import C12Tips from '../demo/12.tips';
import C2_1PictureDisplayCount from '../demo/2.1.pictureDisplayCount';
import C2Picture from '../demo/2.picture';
import C4Multiple from '../demo/4.multiple';
import C5FileListCustomize from '../demo/5.fileListCustomize';
import C6ManualUpload from '../demo/6.manualUpload';
import C7Drap from '../demo/7.drap';
import C8ValidateComplsed from '../demo/8.validate.complsed';
import C8Validate from '../demo/8.validate';
import C9Directory from '../demo/9.directory';
import CCustomizeIconScreenshot from '../demo/CustomizeIcon.screenshot';
import CStatus from '../demo/status';
import CUploadDisabledScreenshot from '../demo/upload-disabled.screenshot';
if (!(C0Basic as Story).storyName) {
  (C0Basic as Story).storyName = '0 basic';
}
if (!(C1Filelist as Story).storyName) {
  (C1Filelist as Story).storyName = '1 filelist';
}
if (!(C10Form as Story).storyName) {
  (C10Form as Story).storyName = '10 form';
}
if (!(C12Tips as Story).storyName) {
  (C12Tips as Story).storyName = '12 tips';
}
if (!(C2_1PictureDisplayCount as Story).storyName) {
  (C2_1PictureDisplayCount as Story).storyName = '2 1 picture display count';
}
if (!(C2Picture as Story).storyName) {
  (C2Picture as Story).storyName = '2 picture';
}
if (!(C4Multiple as Story).storyName) {
  (C4Multiple as Story).storyName = '4 multiple';
}
if (!(C5FileListCustomize as Story).storyName) {
  (C5FileListCustomize as Story).storyName = '5 file list customize';
}
if (!(C6ManualUpload as Story).storyName) {
  (C6ManualUpload as Story).storyName = '6 manual upload';
}
if (!(C7Drap as Story).storyName) {
  (C7Drap as Story).storyName = '7 drap';
}
if (!(C8ValidateComplsed as Story).storyName) {
  (C8ValidateComplsed as Story).storyName = '8 validate complsed';
}
if (!(C8Validate as Story).storyName) {
  (C8Validate as Story).storyName = '8 validate';
}
if (!(C9Directory as Story).storyName) {
  (C9Directory as Story).storyName = '9 directory';
}
if (!(CCustomizeIconScreenshot as Story).storyName) {
  (CCustomizeIconScreenshot as Story).storyName = 'Customize icon screenshot';
}
if (!(CStatus as Story).storyName) {
  (CStatus as Story).storyName = 'Status';
}
if (!(CUploadDisabledScreenshot as Story).storyName) {
  (CUploadDisabledScreenshot as Story).storyName = 'Upload disabled screenshot';
}

export default {
  title: 'Components/Upload/截图测试',
  tags: ['autodocs'],
};

export {
  C0Basic,
  C1Filelist,
  C10Form,
  C12Tips,
  C2_1PictureDisplayCount,
  C2Picture,
  C4Multiple,
  C5FileListCustomize,
  C6ManualUpload,
  C7Drap,
  C8ValidateComplsed,
  C8Validate,
  C9Directory,
  CCustomizeIconScreenshot,
  CStatus,
  CUploadDisabledScreenshot,
};
