// @universe-design/react/src/upload/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0BasicComposed from '../demo/0.basic.composed';
import C1FilelistComposed from '../demo/1.filelist.composed';
import C10FormComposed from '../demo/10.form.composed';
import C11Request from '../demo/11.request';
import C12TipsComposed from '../demo/12.tips.composed';
import C13FileItem from '../demo/13.fileItem';
import C14Dnd from '../demo/14.dnd';
import C15VirtualList from '../demo/15.virtual-list';
import C16NameRenderer from '../demo/16.name-renderer';
import C17Download from '../demo/17.download';
import C2_1PictureDisplayCountComposed from '../demo/2.1.pictureDisplayCount.composed';
import C2PictureComposed from '../demo/2.picture.composed';
import C4MultipleComposed from '../demo/4.multiple.composed';
import C5FileListCustomizeComposed from '../demo/5.fileListCustomize.composed';
import C6ManualUploadComposed from '../demo/6.manualUpload.composed';
import C7DragComposed from '../demo/7.drag.composed';
import C9DirectoryComposed from '../demo/9.directory.composed';
import CHideTrigger from '../demo/hide-trigger';
if (!(C0BasicComposed as Story).storyName) {
  (C0BasicComposed as Story).storyName = '0. 基础用法';
  (C0BasicComposed as Story).parameters = {
    docs: {
      description: {
        story:
          '点击按钮弹出文件选择框，选择文件后自动上传。这里的按钮可以自己根据实际场景换成其他页面元素来触发文件选择。\n\n',
      },
    },
  };
}
if (!(C1FilelistComposed as Story).storyName) {
  (C1FilelistComposed as Story).storyName = '2. 文件列表';
  (C1FilelistComposed as Story).parameters = {
    docs: {
      description: {
        story:
          '组件内置了 4 种文件列表展示形式：\n\n- `<UploadTextFileList />`: 文字列表\n- `<UploadSimpleTextFileList />`: 简略版文字列表\n- `<UploadCardFileList />`: 卡片式列表\n- `<UploadPictureFileList />`: 照片墙\n\n',
      },
    },
  };
}
if (!(C10FormComposed as Story).storyName) {
  (C10FormComposed as Story).storyName = '10. 在Form中使用Upload';
  (C10FormComposed as Story).parameters = {
    docs: {
      description: {
        story: '这里给出一个 Upload 组件和 Form 一起使用的例子：\n\n',
      },
    },
  };
}
if (!(C11Request as Story).storyName) {
  (C11Request as Story).storyName = '11. 自定义请求与请求参数';
  (C11Request as Story).parameters = {
    docs: {
      description: {
        story:
          '### 自定义请求参数\n\n可通过 method、data、headers 自定义上传接口的请求方式、携带数据、自定义请求头部。也可以使用 `customRequest` 完全自定义上传逻辑。下面的例子示意了如何使用 `customRequest`。\n\n',
      },
    },
  };
}
if (!(C12TipsComposed as Story).storyName) {
  (C12TipsComposed as Story).storyName = '12. 提示说明';
  (C12TipsComposed as Story).parameters = {
    docs: {
      description: {
        story: '可在组合使用时，灵活添加自定义区域，如提示说明。\n\n',
      },
    },
  };
}
if (!(C13FileItem as Story).storyName) {
  (C13FileItem as Story).storyName = '13. 使用单文件组件';
  (C13FileItem as Story).parameters = {
    docs: {
      description: {
        story:
          '可使用各种类型的单文件组件。\n**UploadXxxFile Vs XxxFile**\n`TextFile`、`PictureFile`、`CardFile` 这些组件都是纯 UI 组件，仅当需要在 `ComposableUpload` 之外复用样式时使用。\n其他情况都应当使用 `UploadTextFile`、`UploadSimpleTextFile`、`UploadCardFile` 等组件。\n\n',
      },
    },
  };
}
if (!(C14Dnd as Story).storyName) {
  (C14Dnd as Story).storyName = '14. 拖拽排序';
  (C14Dnd as Story).parameters = {
    docs: {
      description: {
        story: '可借助 `react-dnd` 等库，实现拖拽排序。\n\n',
      },
    },
  };
}
if (!(C15VirtualList as Story).storyName) {
  (C15VirtualList as Story).storyName = '15. 虚拟列表';
  (C15VirtualList as Story).parameters = {
    docs: {
      description: {
        story: '当文件数量较大时，可以使用组合用法配合虚拟列表提高渲染性能。\n\n',
      },
    },
  };
}
if (!(C16NameRenderer as Story).storyName) {
  (C16NameRenderer as Story).storyName = '16. 自定义文件名样式';
  (C16NameRenderer as Story).parameters = {
    docs: {
      description: {
        story: '可以对单文件组件自定义文件名样式。\n',
      },
    },
  };
}
if (!(C17Download as Story).storyName) {
  (C17Download as Story).storyName = '17. 下载文件';
  (C17Download as Story).parameters = {
    docs: {
      description: {
        story:
          '`<UploadTextFile />` 支持自定义操作按钮，当传递 `actions` （ReactNode 数组）时，内置按钮将不再展示。\n\n同时，也可以使用 `TextFileAction` 引用内置按钮：\n\n1. `<TextFileAction.Preview />`: 内置预览按钮\n2. `<TextFileAction.Retry />`: 内置重新上传按钮\n3. `<TextFileAction.Remove />`: 内置删除按钮\n4. `<TextFileAction.Download />`: 内置下载按钮\n\n  <code src="./17.download.tsx" />\n',
      },
    },
  };
}
if (!(C2_1PictureDisplayCountComposed as Story).storyName) {
  (C2_1PictureDisplayCountComposed as Story).storyName = '3. 图片墙收起过多图片';
  (C2_1PictureDisplayCountComposed as Story).parameters = {
    docs: {
      description: {
        story:
          'UploadPictureFileList 组件可设置`maxDisplayedPictureCount`属性来显示图片数量，超出此数量的图片将被收起在`+n`里面。\n\n',
      },
    },
  };
}
if (!(C2PictureComposed as Story).storyName) {
  (C2PictureComposed as Story).storyName = '2. 图片墙';
  (C2PictureComposed as Story).parameters = {
    docs: {
      description: {
        story:
          '这里是一个图片墙示例。如果未设置`accept`组件会将其设置为`image/*`，允许所有图片上传。你也可以传入自定义`accept`(用法见 API 说明)。\n\n',
      },
    },
  };
}
if (!(C4MultipleComposed as Story).storyName) {
  (C4MultipleComposed as Story).storyName = '4. 可选择多个文件';
  (C4MultipleComposed as Story).parameters = {
    docs: {
      description: {
        story: '默认每次打开文件选择框，只能选择一个文件，可以添加 multiple 属性支持一次选择多个文件。\n',
      },
    },
  };
}
if (!(C5FileListCustomizeComposed as Story).storyName) {
  (C5FileListCustomizeComposed as Story).storyName = '5. 预览、下载与删除';
  (C5FileListCustomizeComposed as Story).parameters = {
    docs: {
      description: {
        story:
          'Upload 组件默认支持预览、下载、删除、重试功能。\n其中预览和下载需要使用到文件的 url 属性，默认的预览动作为在新标签页打开文件 url，默认的下载动作也是会尝试通过 url 下载文件。\n\n### 自定义操作按钮\n\n可以通过`showPreviewIcon`、`showDownloadIcon`、`showRemoveIcon` 等属性自定义是否支持预览、下载、删除等操作，也可以通过 `previewIcon` 、`downloadIcon` 、 `removeIcon` 等属性自定义操作图标：\n\n了解支持对哪些操作按钮自定义，可参考最后的API说明。\n',
      },
    },
  };
}
if (!(C6ManualUploadComposed as Story).storyName) {
  (C6ManualUploadComposed as Story).storyName = '6. 手动上传';
  (C6ManualUploadComposed as Story).parameters = {
    docs: {
      description: {
        story:
          '当 beforeUpload 钩子返回 false 时，Upload 组件不会上传该文件。\n这可以用来处理选择文件和上传动作不在同时发生，或希望自己操作提交文件过程的场景。\n',
      },
    },
  };
}
if (!(C7DragComposed as Story).storyName) {
  (C7DragComposed as Story).storyName = '7. 拖拽上传';
  (C7DragComposed as Story).parameters = {
    docs: {
      description: {
        story:
          '可设置 `UploadTrigger` 的 type 属性为 `drag` 来允许拖拽上传。可以使用预设的 `UploadDragArea` 组件来设置拖拽区域样式，也可以完全自定义。\n\n\n也可通过 slot 完全自己设置拖拽区域的表现形式。\n',
      },
    },
  };
}
if (!(C9DirectoryComposed as Story).storyName) {
  (C9DirectoryComposed as Story).storyName = '9. 文件夹上传';
  (C9DirectoryComposed as Story).parameters = {
    docs: {
      description: {
        story:
          '支持将整个文件夹上传。该功能浏览器兼容有限，可参考[Directory selection from file input](https://caniuse.com/input-file-directory)\n\n',
      },
    },
  };
}
if (!(CHideTrigger as Story).storyName) {
  (CHideTrigger as Story).storyName = '6. 隐藏上传按钮';
  (CHideTrigger as Story).parameters = {
    docs: {
      description: {
        story:
          '有时我们希望上传完成后隐藏上传按钮。这里给出一个仅允许上传一个文件，并且选择文件之后，隐藏上传按钮的例子：\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Upload/官网示例',
  tags: ['autodocs'],
};

export {
  C0BasicComposed,
  C1FilelistComposed,
  C10FormComposed,
  C11Request,
  C12TipsComposed,
  C13FileItem,
  C14Dnd,
  C15VirtualList,
  C16NameRenderer,
  C17Download,
  C2_1PictureDisplayCountComposed,
  C2PictureComposed,
  C4MultipleComposed,
  C5FileListCustomizeComposed,
  C6ManualUploadComposed,
  C7DragComposed,
  C9DirectoryComposed,
  CHideTrigger,
};
