// @universe-design/react/src/form/utils/validateUtil.ts
import type { RuleItem, ValidateError, ValidateMessages as AsyncValidateMessages } from 'async-validator';
import AsyncValidator from 'async-validator';
import omit from 'lodash/omit';
import cloneDeep from 'lodash/cloneDeep';
import mapValues from 'lodash/mapValues';
import { isNonNullable } from '../../util';
import type { ValidateMessages } from '../interface';
import type { IntlMessage } from '../../shared/locales/types';

type Value = any;

export async function validateRule(
  namePathString: string,
  value: Value,
  rule: RuleItem,
  options: {
    validateMessages: ValidateMessages;
  },
  messageVariables?: Record<string, string>,
): Promise<string[]> {
  // Bug of `async-validator`
  const cloneRule = omit(rule, ['ruleIndex', 'validateTrigger']);

  // We should special handle array validate
  // let subRuleField: Rule | null = null;
  // if (cloneRule && cloneRule.type === 'array' && cloneRule.defaultField) {
  //   subRuleField = cloneRule.defaultField;
  //   delete cloneRule.defaultField;
  // }

  const validator = new AsyncValidator({
    [namePathString]: [cloneRule],
  });

  const messages = options.validateMessages;
  /**
   * 这里故意传递一个不符合类型约束的 messages
   * 为了让 async-validator 不处理 message
   * 后续我们自己处理 message
   */
  validator.messages(cloneDeep(messages as unknown as AsyncValidateMessages));

  let result: (string | IntlMessage | undefined)[] = [];

  try {
    await validator.validate({ [namePathString]: value });
  } catch (_errObj) {
    const errObj = _errObj as {
      errors: (ValidateError & { message: string | IntlMessage })[] | null;
    };

    if (errObj.errors) {
      result = errObj.errors.map(err => err.message);
    } else {
      console.error(errObj);
      result = [messages.default];
    }
  }

  /**
   * Replace message with variables
   * 这里必须将 value 转为字符串，因为可能会包含数组、正则
   **/
  const kv = mapValues(
    {
      ...(rule as Record<string, string | number>),
      name: namePathString,
      enum: (rule.enum || []).join(', '),
      ...messageVariables,
    },
    v => v?.toString(),
  );

  const fillVariableResult = result.filter(isNonNullable).map(error => {
    if (typeof error === 'object' && 'f' in error) {
      return (error.f(kv) as string[]).toString();
    }
    return error;
  });

  return fillVariableResult;
}
