// @universe-design/react/lib/rate/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Rate = void 0;
var _react = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _reactUtilities = require("@universe-design/react-utilities");
var _stylex2 = require("../stylex");
var _hooks = require("../config-provider/hooks");
var _tooltip = require("../tooltip");
var _CollectFilled = _interopRequireDefault(require("@universe-design/icons-react/CollectFilled"));
var _util = require("../util");
var _noop = _interopRequireDefault(require("lodash/noop"));
require("./style");
const Rate = props => {
  const {
    style,
    className,
    character,
    allowHalf,
    allowClear = true,
    value: valueInProps,
    defaultValue = 0,
    readonly = false,
    count = 5,
    tooltips = [],
    showDescriptions = true,
    descriptions = [],
    autoFocus = false,
    onHoverChange,
    onChange
  } = props;
  const cssScope = (0, _hooks.useCssScope)('rate');
  const cls = (0, _clsx.default)(cssScope(), className);
  const {
    components: {
      rate: rateLocale
    }
  } = (0, _hooks.useLocale)();
  const rateRef = (0, _react.useRef)(null);
  const [value, setValue] = (0, _reactUtilities.useControlledState)(defaultValue, valueInProps, {
    onChange
  });
  const [hoverValue, setHoverValue] = (0, _react.useState)();
  const [cleanedValue, setCleanedValue] = (0, _react.useState)();
  (0, _react.useEffect)(() => {
    if (rateRef.current && autoFocus) {
      rateRef.current.focus();
    }
  }, [rateRef, autoFocus]);
  const getValue = (index, e) => {
    let currentValue = index + 1;
    const {
      left,
      width
    } = e.currentTarget.getBoundingClientRect();
    if (allowHalf && e.clientX - left < width / 2) {
      // eslint-disable-next-line @typescript-eslint/no-magic-numbers
      currentValue -= 0.5;
    }
    return currentValue;
  };
  const calculateDes = () => {
    const index = parseInt((hoverValue || value || 0).toFixed(), 10);
    return descriptions[index] || rateLocale.descriptions[index];
  };
  const handleMouseMove = (index, e) => {
    const currentValue = getValue(index, e);
    // 避免点击触发清除后，再次设置选中态
    /* istanbul ignore else  */
    if (currentValue !== cleanedValue) {
      setHoverValue(currentValue);
      (0, _util.isFunction)(onHoverChange) && onHoverChange(currentValue);
    }
  };
  const handleChoose = (index, e) => {
    const currentValue = getValue(index, e);
    const needReset = allowClear && currentValue === value;
    setValue(needReset ? 0 : currentValue);
    setHoverValue(undefined);
    setCleanedValue(needReset ? currentValue : undefined);
  };
  const wrapTooltips = (index, node) => {
    const tip = tooltips[index];
    if (tip) {
      return /*#__PURE__*/_react.default.createElement(_tooltip.Tooltip, {
        key: `rate-tooltip-${index}`,
        title: tip
      }, node);
    }
    return node;
  };
  const renderCharacterEl = index => {
    if ((0, _util.isFunction)(character)) {
      return character({
        ...props,
        index
      });
    } else {
      return character || /*#__PURE__*/_react.default.createElement(_CollectFilled.default, null);
    }
  };
  const renderCharacter = index => {
    let preciseStyle;
    const currentIndex = index + 1;
    const currentValue = hoverValue || value;
    const includes = currentValue && currentValue > index && currentValue < currentIndex;
    const hasHalf = allowHalf ? includes : false;
    const itemCls = (0, _clsx.default)(cssScope('__character'), {
      [cssScope('__character--full')]: currentValue && currentValue >= currentIndex,
      [cssScope('__character--half')]: readonly && includes || hasHalf
    });
    if (readonly && currentValue && includes) {
      preciseStyle = {
        width: `${(currentValue - index) * 100}%`
      };
    }
    return wrapTooltips(index, /*#__PURE__*/_react.default.createElement("div", {
      className: itemCls,
      key: index,
      onMouseMove: readonly ? _noop.default : e => handleMouseMove(index, e),
      onClick: readonly ? _noop.default : e => handleChoose(index, e)
    }, /*#__PURE__*/_react.default.createElement("div", {
      style: preciseStyle,
      className: cssScope('__character-half-body')
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: cssScope('__character-half-body-wrap')
    }, renderCharacterEl(index))), /*#__PURE__*/_react.default.createElement("div", {
      className: cssScope('__character-body')
    }, renderCharacterEl(index))));
  };
  return /*#__PURE__*/_react.default.createElement("div", {
    ref: rateRef,
    style: style,
    className: cls,
    onMouseLeave: () => {
      setHoverValue(undefined);
      setCleanedValue(undefined);
      (0, _util.isFunction)(onHoverChange) && onHoverChange(0);
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: cssScope('__characters')
  }, (() => {
    const items = [];
    for (let i = 0; i < count; i++) {
      items.push(renderCharacter(i));
    }
    return items;
  })()), (() => {
    if (showDescriptions) {
      return /*#__PURE__*/_react.default.createElement("div", {
        className: (0, _clsx.default)(cssScope('__des'), (0, _stylex.default)(_stylex2.cssTypographyBody0))
      }, calculateDes());
    }
    return null;
  })());
};
exports.Rate = Rate;
