// @universe-design/react/lib/select/common.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultFilterOption = exports.UD_SELECT_OPTION_ID = exports.UD_SELECT_OPTION_GROUP_ID = void 0;
exports.getOption = getOption;
exports.getReactKey = getReactKey;
exports.isSelectOption = isSelectOption;
exports.isSelectOptionGroup = isSelectOptionGroup;
exports.useSelectListCssScope = useSelectListCssScope;
exports.useSelectListGroupCssScope = useSelectListGroupCssScope;
exports.useSelectListItemCssScope = useSelectListItemCssScope;
var _escapeRegExp = _interopRequireDefault(require("lodash/escapeRegExp"));
var _hooks = require("../config-provider/hooks");
function useSelectListCssScope() {
  return (0, _hooks.useCssScope)('select__list');
}
function useSelectListItemCssScope() {
  return (0, _hooks.useCssScope)('select__list__item');
}
function useSelectListGroupCssScope() {
  return (0, _hooks.useCssScope)('select__list__group');
}

/**
 * ======== 判断是不是我们提供的 Option 和 OptionGroup 组件 ========
 */
const UD_SELECT_OPTION_ID = exports.UD_SELECT_OPTION_ID = 'UD_SELECT_OPTION';
const UD_SELECT_OPTION_GROUP_ID = exports.UD_SELECT_OPTION_GROUP_ID = 'UD_SELECT_OPTION_GROUP';
function isSelectOption(node) {
  var _node$type;
  return node === null || node === void 0 ? void 0 : (_node$type = node.type) === null || _node$type === void 0 ? void 0 : _node$type[UD_SELECT_OPTION_ID];
}
function isSelectOptionGroup(node) {
  var _node$type2;
  return node === null || node === void 0 ? void 0 : (_node$type2 = node.type) === null || _node$type2 === void 0 ? void 0 : _node$type2[UD_SELECT_OPTION_GROUP_ID];
}
const defaultFilterOption = (inputValue, option) => {
  const keyword = typeof option.label === 'string' ? option.label : String(option.value);
  return !inputValue || new RegExp((0, _escapeRegExp.default)(inputValue), 'gi').test(keyword);
};
exports.defaultFilterOption = defaultFilterOption;
function getOption(value, optionMap, selectedMap) {
  var _ref, _optionMap$get;
  // FIXME 如果没有在 option map 中找到的话做一个假的 option 返回
  return (_ref = (_optionMap$get = optionMap.get(value)) !== null && _optionMap$get !== void 0 ? _optionMap$get : selectedMap.get(value)) !== null && _ref !== void 0 ? _ref : {
    value
  };
}
function getReactKey(value) {
  return typeof value === 'boolean' ? String(`${value}-bool`) : value;
}
