// @universe-design/react/lib/select/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  Select: true,
  SelectContext: true,
  SelectSearchAction: true,
  SelectActionType: true,
  useSelectInputValue: true
};
exports.Select = void 0;
Object.defineProperty(exports, "SelectActionType", {
  enumerable: true,
  get: function () {
    return _internal.SelectActionType;
  }
});
Object.defineProperty(exports, "SelectContext", {
  enumerable: true,
  get: function () {
    return _SelectContext.SelectContext;
  }
});
Object.defineProperty(exports, "SelectSearchAction", {
  enumerable: true,
  get: function () {
    return _internal.SelectSearchAction;
  }
});
Object.defineProperty(exports, "useSelectInputValue", {
  enumerable: true,
  get: function () {
    return _useSelectInputValue.useSelectInputValue;
  }
});
var _configProvider = require("../config-provider");
var _Select2 = require("./Select");
var _SelectAsync = require("./SelectAsync");
var _SelectOption = require("./SelectOption");
require("./style");
var _selectGenerator = require("./selectGenerator");
Object.keys(_selectGenerator).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _selectGenerator[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _selectGenerator[key];
    }
  });
});
var _SelectContext = require("./context/SelectContext");
var _interface = require("./interface");
Object.keys(_interface).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interface[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interface[key];
    }
  });
});
var _internal = require("./internal.interface");
var _useSelectInputValue = require("./context/useSelectInputValue");
const Select = exports.Select = Object.assign(_Select2.Select, {
  Option: _SelectOption.SelectOption,
  OptGroup: _SelectOption.SelectOptionGroup,
  Async: _SelectAsync.SelectAsync,
  [_configProvider.COMPONENT_VERSION_ID]: "3.28.3"
});
