// @universe-design/react/lib/select/selectGenerator.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.selectGenerator = selectGenerator;
var _DownBoldOutlined = _interopRequireDefault(require("@universe-design/icons-react/DownBoldOutlined"));
var _EmptyFilled = _interopRequireDefault(require("@universe-design/icons-react/EmptyFilled"));
var _clsx = _interopRequireDefault(require("clsx"));
var React = _interopRequireWildcard(require("react"));
var _hooks = require("../config-provider/hooks");
var _usePopperSidecar = require("../config-provider/hooks/usePopperSidecar");
var _hooks2 = require("../hooks");
var _useComplementEvent = require("../hooks/useComplementEvent");
var _SlideUpDownTransition = require("../shared/components/transition/SlideUpDownTransition");
var _util = require("../util");
var _composeRefs = _interopRequireDefault(require("../util/composeRefs"));
var _Selector = require("./Selector");
var _SelectorTag = require("./SelectorTag");
var _MultipleSelectorContext = require("./context/MultipleSelectorContext");
var _SelectContext = require("./context/SelectContext");
var _useSelection = require("./context/useSelection");
require("./style");
var _reactProvider = require("@universe-design/react-provider");
/* eslint-disable max-lines-per-function */

function selectGenerator(config) {
  const InternalSelect = /*#__PURE__*/React.forwardRef((props, ref) => {
    const locale = (0, _hooks.useLocale)();
    const ctxSize = (0, _hooks.useComponentSize)();
    const inputGroupSize = (0, _hooks.useInputGroupSize)();
    const ctxGetPopupContainer = (0, _reactProvider.usePopupPortalContainer)();
    const {
      placeholder = locale.global.placeholder,
      allowClear = true,
      style,
      className,
      size = inputGroupSize || ctxSize,
      bordered = 'normal',
      dropdownMatchSelectWidth = false,
      error = false,
      maxTagCount,
      overlayNode,
      getPopupContainer = ctxGetPopupContainer,
      showArrow = true,
      maxTagRenderer = _SelectorTag.defaultMaxTagRenderer,
      clearIcon = /*#__PURE__*/React.createElement(_EmptyFilled.default, null),
      suffixIcon = /*#__PURE__*/React.createElement(_DownBoldOutlined.default, null),
      dropdownClassName,
      dropdownStyle,
      triggerElement,
      onMouseEnter,
      onMouseLeave,
      multipleSelectorMaxHeight = 188,
      optionFilterProp,
      removeIcon,
      placement = 'bottom',
      forceRender,
      destroyOnHide = false
    } = props;
    process.env.NODE_ENV !== "production" ? (0, _util.warning)(!optionFilterProp && !removeIcon, (0, _util.compatibleDevWarning)(InternalSelect.displayName || InternalSelect.name, ['optionFilterProp', 'removeIcon'])) : void 0;
    const triggerElementEnable = (0, _util.isNonNullable)(triggerElement);
    const [overlayElement, overlayRef] = React.useState(null);
    // selectorRef
    const [selectorContainerRef, {
      width: selectorContainerWidth
    }] = (0, _hooks2.useSize)();
    const [selectorContainerElement, selectorContainerStateRef] = React.useState(null);
    const selectCtx = (0, _useSelection.useSelection)({
      ...props,
      size
    }, config);
    const {
      focusState,
      overlayVisible,
      focus,
      blur,
      setOverlayVisible,
      dropdownCssScope,
      cssScope,
      inputRef,
      cancelDelayCloseOverlay,
      // onInputFocus,
      setInputting
    } = selectCtx;

    /**
     * 控制 select popup 在点击非内部元素时自动关闭
     */
    (0, _useComplementEvent.useComplementEvent)('mousedown', {
      listener: React.useCallback(() => {
        console.log('going to close select popup');
        setOverlayVisible(false);
        // For jest test, in browser input should blur when use click other places than inputElement
        blur();
      }, [blur, setOverlayVisible]),
      // 第一次打开 overlay, window mousedown 触发之前 overlayRef 内容还是空的，所以这里需要使用 ref 来保证正确获取 dom
      elements: React.useMemo(() => [overlayElement, selectorContainerElement], [overlayElement, selectorContainerElement]),
      skipped: !overlayVisible,
      options: {
        /**
         * 防止 mousedown 的元素也监听了 mousedown 且阻止了冒泡
         */
        capture: true
      }
    });
    React.useImperativeHandle(ref, () => ({
      focus,
      blur,
      get input() {
        return inputRef.current;
      }
    }), [blur, focus, inputRef]);

    // ============================= Popup ==============================
    const popupStyle = {
      minWidth: selectorContainerWidth,
      ...(dropdownStyle !== null && dropdownStyle !== void 0 ? dropdownStyle : {})
    };
    if (typeof dropdownMatchSelectWidth === 'number') {
      popupStyle.width = dropdownMatchSelectWidth;
    } else if (dropdownMatchSelectWidth) {
      popupStyle.width = selectorContainerWidth;
    }
    const rootCls = (0, _clsx.default)(cssScope(), className, {
      [cssScope('--focus')]: focusState
    });
    const direction = (0, _hooks.useDirection)();
    const Popper = (0, _usePopperSidecar.usePopperSidecar)();
    let selectorNode = /*#__PURE__*/React.createElement(_Selector.Selector, {
      showArrow: showArrow,
      allowClear: allowClear,
      maxTagCount: maxTagCount,
      bordered: bordered,
      size: size,
      placeholder: placeholder,
      error: error,
      maxTagRenderer: maxTagRenderer,
      clearIcon: clearIcon,
      suffixIcon: suffixIcon
    });
    if (triggerElementEnable) {
      selectorNode = (0, _util.isFunction)(triggerElement) ? triggerElement(selectCtx) : triggerElement !== null && triggerElement !== void 0 ? triggerElement : /*#__PURE__*/React.createElement("div", null);
    }
    const activeTimeoutIdsRef = React.useRef([]);
    React.useEffect(() => () => {
      const activeTimeoutIds = activeTimeoutIdsRef.current;
      activeTimeoutIds.forEach(timeoutId => clearTimeout(timeoutId));
      activeTimeoutIds.splice(0, activeTimeoutIds.length);
    }, []);
    const popupPlacement = React.useMemo(() => {
      if (['top', 'bottom'].includes(placement)) {
        return direction === 'rtl' ? `${placement}Right` : `${placement}Left`; // 目前 right 的 css 样式还没有加
      }
      throw new Error(`Placement ${placement} is not defined.`);
    }, [direction, placement]);
    return /*#__PURE__*/React.createElement("div", {
      ref: (0, _composeRefs.default)(selectorContainerRef, selectorContainerStateRef),
      className: rootCls,
      style: style,
      onMouseEnter: onMouseEnter,
      onMouseLeave: onMouseLeave,
      onKeyDown: selectCtx.onInternalKeyDown,
      onKeyUp: selectCtx.onInternalKeyUp,
      onBlur: selectCtx.onContainerBlur,
      onFocus: selectCtx.onContainerFocus,
      onMouseDown: event => {
        setInputting(true);
        const activeTimeoutIds = activeTimeoutIdsRef.current;
        const {
          target
        } = event;
        const popupElement = overlayElement;

        // We should give focus back to selector if clicked item is not focusable
        // 这段代码是为了保证 dropdown 里面有 input 的时候在点击 input 和失焦 input 时都可以正常工作
        if (popupElement && popupElement.contains(target)) {
          const timeoutId = window.setTimeout(() => {
            const index = activeTimeoutIds.indexOf(timeoutId);
            if (index !== -1) {
              // 进入处理阶段，把 timeout id 删掉
              activeTimeoutIds.splice(index, 1);
            }

            // 取消延迟执行的 blur
            cancelDelayCloseOverlay();

            // 如果当前 focus 的不是 popup 内部的元素，就聚焦 selector input
            if (!popupElement.contains(document.activeElement)) {
              var _inputRef$current;
              (_inputRef$current = inputRef.current) === null || _inputRef$current === void 0 ? void 0 : _inputRef$current.focus();
            }
          });
          activeTimeoutIds.push(timeoutId);
        }
      }
    }, /*#__PURE__*/React.createElement(_MultipleSelectorContext.MultipleSelectorContextProvider, {
      value: React.useMemo(() => ({
        maxHeight: multipleSelectorMaxHeight
      }), [multipleSelectorMaxHeight])
    }, /*#__PURE__*/React.createElement(_SelectContext.SelectContext.Provider, {
      value: selectCtx
    }, /*#__PURE__*/React.createElement(Popper, {
      getPopupContainer: getPopupContainer,
      forceRender: forceRender,
      destroyOnHide: destroyOnHide,
      action: undefined,
      showAction: [],
      hideAction: [],
      popupPlacement: popupPlacement,
      popupMotion: "slide-up",
      prefixCls: dropdownCssScope(),
      popupClassName: dropdownClassName,
      visible: overlayVisible,
      popupStyle: popupStyle,
      popup: /*#__PURE__*/React.createElement("div", {
        ref: overlayRef
      }, overlayNode),
      enableRestoreFocus: false,
      initialFocus: null,
      showArrow: false,
      arrowHasStroke: false,
      Transition: _SlideUpDownTransition.SlideUpDownTransition,
      enableMaxHeight: false,
      defaultMainAxisOffsetWithoutArrow: 4
    }, selectorNode))));
  });
  if (process.env.NODE_ENV !== "production") {
    InternalSelect.displayName = 'InternalSelect';
  }
  return InternalSelect;
}
