// @universe-design/react/lib/select/SelectorTag.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultTagRenderer = exports.defaultMaxTagRenderer = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _clsx = _interopRequireDefault(require("clsx"));
var _react = _interopRequireWildcard(require("react"));
var _tag = require("../tag");
var _SelectContext = require("./context/SelectContext");
var _popover = require("../popover");
var _space = require("../space");
/**
 * 默认的 tag renderer
 */
function MultipleSelectTag(_ref) {
  let {
    value,
    title,
    closable,
    disabled,
    onClose,
    children
  } = _ref;
  const {
    selectorCssScope: cssScope,
    size
  } = (0, _react.useContext)(_SelectContext.SelectContext);
  const className = (0, _clsx.default)(cssScope('__tag'));
  return /*#__PURE__*/_react.default.createElement(_tag.Tag, {
    key: value,
    title: title,
    className: className,
    color: "neutral-option",
    size: size,
    closable: closable,
    visible: true,
    onClose: onClose,
    disabled: disabled
  }, children);
}
const defaultTagRenderer = props => /*#__PURE__*/_react.default.createElement(MultipleSelectTag, props);

/**
 * 默认的 max tag renderer
 */
exports.defaultTagRenderer = defaultTagRenderer;
const MultipleSelectMaxTag = _ref2 => {
  let {
    disabled,
    tags
  } = _ref2;
  const {
    selectorCssScope: cssScope,
    size,
    tagRenderer
  } = (0, _react.useContext)(_SelectContext.SelectContext);
  const className = (0, _clsx.default)(cssScope('__tag'));
  const tagNode = /*#__PURE__*/_react.default.createElement(_tag.Tag, {
    className: className,
    style: {
      justifyContent: 'center'
    },
    color: "neutral-option",
    size: size,
    disabled: disabled
  }, `+${tags.length}`);
  return /*#__PURE__*/_react.default.createElement(_popover.Popover, {
    mode: "pure",
    overlayClassName: cssScope('__overlay'),
    content: /*#__PURE__*/_react.default.createElement("div", {
      className: cssScope('__overlay-scrollArea')
    }, /*#__PURE__*/_react.default.createElement(_space.Space, {
      size: 4,
      wrap: true,
      inline: false,
      shouldItemConfinedToContainer: true,
      itemKey: index => tags[index].value
    }, tags.map(tag => /*#__PURE__*/_react.default.createElement(_react.default.Fragment, {
      key: tag.value
    }, tagRenderer(tag))))),
    visible: tags.length === 0 ? false : undefined // 不指定这个在 tags 从有到无的时候会显示一个空的 popover
    ,
    disabled: tags.length === 0
  }, tagNode);
};
const defaultMaxTagRenderer = (props, tags) => /*#__PURE__*/_react.default.createElement(MultipleSelectMaxTag, (0, _extends2.default)({}, props, {
  tags: tags
}));
exports.defaultMaxTagRenderer = defaultMaxTagRenderer;
