// @universe-design/react/lib/select/SelectOverlay.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SelectOverlay = void 0;
exports.getNextActiveOption = getNextActiveOption;
exports.useSelectOverlay = useSelectOverlay;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _noop = _interopRequireDefault(require("lodash/noop"));
var React = _interopRequireWildcard(require("react"));
var _rcVirtualList = _interopRequireDefault(require("@universe-design/rc-virtual-list"));
var _util = require("../util");
var _reactUtilities = require("@universe-design/react-utilities");
var _common = require("./common");
var _SelectContext = require("./context/SelectContext");
var _SelectOption = require("./SelectOption");
var _hooks = require("../hooks");
var _internal = require("./internal.interface");
var _composeRefs = _interopRequireDefault(require("../util/composeRefs"));
var _list = require("../shared/components/list");
var _checkbox = require("../checkbox");
var _useSelectAll = _interopRequireDefault(require("./context/useSelectAll"));
var _hooks2 = require("../config-provider/hooks");
var _SelectOverlayContainer = require("./SelectOverlayContainer");
// 这里和 option shape 做一下区分

function isGroupFlatternOption(opt) {
  return opt.group;
}
function isSimpleFlatternOption(opt) {
  return !opt.group;
}
function getFirstOption(options) {
  let predicate = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : () => true;
  for (const opt of options) {
    if (isSimpleFlatternOption(opt) && !opt.option.disabled && predicate(opt)) {
      return opt.option;
    }
  }
}
function getLastOption(options) {
  const opt = options.filter(isSimpleFlatternOption).filter(o => !o.option.disabled).pop();
  if (!opt) {
    return undefined;
  }
  return opt.option;
}
function getNextActiveOption(activeOption, options, forward) {
  const curActiveKeyIndex = options.findIndex(opt => isSimpleFlatternOption(opt) && !opt.option.disabled && opt.option.value === (activeOption === null || activeOption === void 0 ? void 0 : activeOption.value));
  if (curActiveKeyIndex === -1) {
    if (forward) {
      return getFirstOption(options);
    } else {
      return getLastOption(options);
    }
  }
  for (let i = curActiveKeyIndex, count = 0; count < options.length; count += 1,
  // eslint-disable-next-line no-nested-ternary
  i = forward // 正向遍历或反向遍历整个数组
  ? i + 1 > options.length - 1 ? 0 : i + 1 : i - 1 < 0 ? options.length - 1 : i - 1) {
    if (i === curActiveKeyIndex) {
      continue;
    }
    const opt = options[i];
    if (isSimpleFlatternOption(opt) && !opt.option.disabled) {
      return opt.option;
    }
  }
}
function useSelectOverlay(props) {
  const {
    inputValue: ctxInputValue,
    overlayVisible,
    valueSet,
    toggle,
    listRef,
    compositionStatusRef,
    set,
    disabled
  } = React.useContext(_SelectContext.SelectContext);
  const {
    defaultActiveFirstOption = true,
    options,
    filterOption,
    filterSort,
    debounceTime,
    onActiveOptionChange
  } = props;
  const [inputValue] = (0, _hooks.useDebounce)(ctxInputValue, debounceTime);

  // =========================== List ===========================
  const flattenOptions = React.useMemo(() => {
    const _filterOption = (_inputValue, _optProps) => {
      if (filterOption === true) {
        return (0, _common.defaultFilterOption)(_inputValue, _optProps);
      } else if (filterOption === false) {
        return true;
      }
      return filterOption(_inputValue, _optProps);
    };
    const shouldSort = filterSort && inputValue;

    /**
     * 只筛选，保留树形结构
     */
    const filteredOptions = options.map(optProps => {
      if (!('options' in optProps) && !_filterOption(inputValue, optProps)) {
        return null;
      }
      if ('options' in optProps) {
        const _filteredOptions = optProps.options.filter(_optProps => _filterOption(inputValue, _optProps));
        if (_filteredOptions.length) {
          return {
            ...optProps,
            options: shouldSort ? _filteredOptions.sort(filterSort) : _filteredOptions
          };
        }
        return null;
      }
      return optProps;
    }).filter(_util.isNonNullable);
    if (shouldSort) {
      filteredOptions.sort(filterSort);
    }

    /**
     * 打平树形结构
     */
    const res = [];
    for (const opt of filteredOptions) {
      if ('options' in opt) {
        res.push({
          group: true,
          option: opt
        });
        for (const _opt of opt.options) {
          res.push({
            group: false,
            option: _opt
          });
        }
      } else {
        res.push({
          group: false,
          option: opt
        });
      }
    }
    return res;
  }, [filterOption, filterSort, inputValue, options]);
  const latestFlattenOptionsRef = (0, _hooks.useLatest)(flattenOptions);
  const {
    onToggleSelectAll,
    hasSelectedAll,
    isSelectAllDisabled,
    selectAllCount
  } = (0, _useSelectAll.default)({
    set,
    flattenOptions,
    valueSet
  });
  // =========================== List ===========================
  const rcVirtualListRef = React.useRef(null);

  /**
   * 滚动到特定选项，支持按照三种方式滚动：
   *  1. index: 根据选项所在 index
   *  2. option value
   *  3. groupKey
   */
  const scrollIntoView = (reference, scrollConfig) => {
    const list = rcVirtualListRef.current;
    if (!list) {
      return;
    }
    let index = -1;
    if (typeof reference === 'number') {
      index = reference;
    } else if (reference.type === 'option') {
      index = flattenOptions.findIndex(flattenOption => !flattenOption.group && flattenOption.option.value === reference.value);
    } else if (reference.type === 'group') {
      index = flattenOptions.findIndex(flattenOption => flattenOption.group && flattenOption.option.key === reference.key);
    }
    if (index !== -1) {
      list.scrollTo({
        index,
        ...scrollConfig
      });
    }
  };

  // ========================== Active ==========================
  /**
   * 逻辑
   * - 优先级
   *   - active 上次关闭时的项
   *   - active 第一个选中项
   *   - active 第一个选项
   *   - defaultActiveFirstOption false
   * - 场景
   *   - 首次打开
   *   - 再次打开
   *   - 选项列表变化，比如搜索，或异步选取等
   */
  const firstOption = React.useMemo(() => getFirstOption(flattenOptions), [flattenOptions]);
  const latestFirstOptionRef = (0, _hooks.useLatest)(firstOption);
  const firstSelectedOption = React.useMemo(() => getFirstOption(flattenOptions, opt => valueSet.has(opt.option.value)), [flattenOptions, valueSet]);
  const defaultOption = React.useMemo(() => {
    if (!defaultActiveFirstOption) {
      return undefined;
    }
    return firstSelectedOption !== null && firstSelectedOption !== void 0 ? firstSelectedOption : firstOption;
  }, [defaultActiveFirstOption, firstOption, firstSelectedOption]);
  const [activeOption, _setActiveOption] = React.useState(defaultOption);
  const setActiveOption = React.useCallback(meta => {
    var _meta$option;
    onActiveOptionChange && onActiveOptionChange((_meta$option = meta.option) === null || _meta$option === void 0 ? void 0 : _meta$option.value, meta);
    _setActiveOption(meta.option);
  }, [onActiveOptionChange]);
  const latestSetActiveOption = (0, _hooks.useEventCallback)(setActiveOption);

  // 初次进入滑到对应的元素
  (0, _hooks.useEffectOnce)(() => {
    scrollIntoView(flattenOptions.findIndex(o => isSimpleFlatternOption(o) && o.option.value === (activeOption === null || activeOption === void 0 ? void 0 : activeOption.value)));
  });

  // ========================= Keyboard =========================
  React.useImperativeHandle((0, _composeRefs.default)(listRef, props.ref), () => ({
    onKeyDown: e => {
      if (e.isDefaultPrevented()) {
        return;
      }
      let up;
      function setActiveAndScrollTo() {
        const nextOption = getNextActiveOption(activeOption, flattenOptions, !up);
        if (nextOption) {
          // 跟随 active 元素滚动
          setActiveOption({
            option: nextOption,
            action: up ? _internal.SelectActiveOptionChangeAction.ArrowUp : _internal.SelectActiveOptionChangeAction.ArrowDown
          });
          scrollIntoView(flattenOptions.findIndex(o => isSimpleFlatternOption(o) && o.option.value === nextOption.value));
        }
      }
      switch (e.key) {
        case _reactUtilities.KeyboardKey.Down: // IE/Edge specific value
        case _reactUtilities.KeyboardKey.ArrowDown:
          up = false;
          setActiveAndScrollTo();
          break;
        case _reactUtilities.KeyboardKey.Up: // IE/Edge specific value
        case _reactUtilities.KeyboardKey.ArrowUp:
          up = true;
          setActiveAndScrollTo();
          break;

        // >>> Select
        case _reactUtilities.KeyboardKey.Enter:
          // overlay 操作选中值或 toggle overlay
          if (activeOption && overlayVisible && !compositionStatusRef.current) {
            e.preventDefault();
            toggle(activeOption.value);
          }
          break;
        default:
      }
    },
    onKeyUp: _noop.default,
    scrollTo: scrollIntoView,
    scrollIntoView,
    setActiveOption: value => {
      var _flattenOptions$filte;
      setActiveOption({
        option: (_flattenOptions$filte = flattenOptions.filter(isSimpleFlatternOption).find(o => o.option.value === value)) === null || _flattenOptions$filte === void 0 ? void 0 : _flattenOptions$filte.option,
        action: _internal.SelectActiveOptionChangeAction.Manual
      });
    }
  }));
  const isFirstMount = React.useRef(true);
  React.useEffect(() => {
    if (isFirstMount.current) {
      isFirstMount.current = false;
      return;
    }
    // 这里限制只有 inputValue 变了才触发重新设置 active option。因为 option list 还可能因为操作有新增和删除，这时候变化 active，可能有超出预期的情况。规范上最明确的预期就是搜索后，尽量保证有 active 选项，排除其他更高优逻辑，就是首个选项 active 兜底，所以限定在搜索时触发就行了 -- @qiuchen.1001

    latestSetActiveOption({
      option: getFirstOption(latestFlattenOptionsRef.current, opt => {
        var _latestFirstOptionRef;
        return opt.option.value === ((_latestFirstOptionRef = latestFirstOptionRef.current) === null || _latestFirstOptionRef === void 0 ? void 0 : _latestFirstOptionRef.value);
      }),
      action: _internal.SelectActiveOptionChangeAction.AfterSearch
    });
  }, [latestFirstOptionRef, latestFlattenOptionsRef, inputValue, latestSetActiveOption]);
  return {
    activeOption,
    setActiveOption,
    rcVirtualListRef,
    flattenOptions,
    onToggleSelectAll,
    hasSelectedAll,
    isSelectAllDisabled,
    selectAllCount,
    disabled
  };
}
const SelectOverlay = exports.SelectOverlay = /*#__PURE__*/React.forwardRef(function (props, ref) {
  const {
    options,
    filterOption,
    defaultActiveFirstOption,
    menuItemSelectedIcon,
    listHeight,
    onPopupScroll,
    itemHeight,
    virtual,
    onMouseEnter,
    filterSort,
    debounceTime = 0,
    showSelectAll,
    onActiveOptionChange
  } = props;
  const cssScope = (0, _common.useSelectListCssScope)();
  const {
    activeOption,
    setActiveOption,
    rcVirtualListRef,
    flattenOptions,
    hasSelectedAll,
    onToggleSelectAll,
    isSelectAllDisabled,
    selectAllCount,
    disabled
  } = useSelectOverlay({
    ref,
    options,
    filterOption,
    defaultActiveFirstOption,
    filterSort,
    debounceTime,
    onActiveOptionChange
  });
  const local = (0, _hooks2.useLocale)();
  // ========================== Render ==========================
  const menuNode = /*#__PURE__*/React.createElement("div", {
    className: cssScope('__menu')
  }, /*#__PURE__*/React.createElement(_rcVirtualList.default, {
    itemKey: option => isGroupFlatternOption(option) ? `group-${option.option.key}` : (0, _common.getReactKey)(option.option.value),
    ref: rcVirtualListRef,
    data: flattenOptions,
    height: listHeight - 3 * 2 - 1 * 2,
    itemHeight: itemHeight,
    fullHeight: false,
    onScroll: onPopupScroll,
    virtual: virtual,
    onMouseEnter: onMouseEnter
  }, (option, itemIndex) => {
    // Group
    if (isGroupFlatternOption(option)) {
      return /*#__PURE__*/React.createElement(_SelectOption.InternalSelectOptionGroup, (0, _extends2.default)({}, option.option, {
        isFirst: itemIndex === 0
      }));
    }

    // Option
    const {
      option: origOpt
    } = option;
    return (
      /*#__PURE__*/
      // 这里加一个空 div 是为了把内部带 margin 的元素包住，让 virtual list 能正常计算，tree，select，cascader
      React.createElement("div", null, /*#__PURE__*/React.createElement(_SelectOption.InternalSelectOption, (0, _extends2.default)({
        active: origOpt.value === (activeOption === null || activeOption === void 0 ? void 0 : activeOption.value),
        key: (0, _common.getReactKey)(origOpt.value),
        menuItemSelectedIcon: menuItemSelectedIcon
      }, origOpt, {
        onMouseMove: () => setActiveOption({
          option: origOpt,
          action: _internal.SelectActiveOptionChangeAction.MouseMove
        })
      })))
    );
  }), showSelectAll && /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(_list.ListDivider, {
    fluid: true
  }), /*#__PURE__*/React.createElement(_list.ListItemLike, null, /*#__PURE__*/React.createElement(_checkbox.Checkbox, {
    disabled: disabled || isSelectAllDisabled,
    checked: hasSelectedAll,
    onChange: onToggleSelectAll
  }, local.global.selectAll, " (", selectAllCount, ")"))));
  return /*#__PURE__*/React.createElement(_SelectOverlayContainer.SelectOverlayContainer, {
    optionsLength: options.length,
    filteredOptionsLength: flattenOptions.length,
    menu: menuNode
  });
});
