// @universe-design/react/lib/select/SelectOption.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SelectOptionGroup = exports.SelectOption = exports.InternalSelectOptionGroup = exports.InternalSelectOption = void 0;
var _urmav4p = _interopRequireDefault(require("./../.stylex/urmav4p"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _clsx = _interopRequireDefault(require("clsx"));
var _react = _interopRequireWildcard(require("react"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _configProvider = require("../config-provider");
var _shared = require("../shared");
var _common = require("./common");
var _SelectContext = require("./context/SelectContext");
var _stylex2 = require("../stylex");
var _Text = require("../text/Text");
//.u10tiqix{min-width:0px;}
/**
 * For users
 */
const SelectOption = () => /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null);
exports.SelectOption = SelectOption;
Object.assign(SelectOption, {
  [_common.UD_SELECT_OPTION_ID]: true,
  [_configProvider.COMPONENT_VERSION_ID]: "3.28.3"
});

/**
 * For internal use
 */
const InternalSelectOption = exports.InternalSelectOption = /*#__PURE__*/(0, _react.forwardRef)((_ref, ref) => {
  let {
    value,
    label,
    labelRenderer,
    disabled = false,
    menuItemSelectedIcon,
    title,
    active,
    className,
    // omit
    children,
    ...restProps
  } = _ref;
  const cssScope = (0, _common.useSelectListItemCssScope)();
  const {
    valueSet,
    toggle,
    inputValue,
    textOverflow
  } = _react.default.useContext(_SelectContext.SelectContext);
  function renderHighligherNode(_value) {
    return /*#__PURE__*/_react.default.createElement(_shared.Highlighter, {
      searchWords: [inputValue],
      textToHighlight: _value,
      markRenderer: text => /*#__PURE__*/_react.default.createElement("span", {
        className: (0, _clsx.default)(!disabled && cssScope('__content-highlight'))
      }, text)
    });
  }
  const selected = valueSet.has(value);
  let labelNode;
  if (labelRenderer) {
    labelNode = labelRenderer(inputValue);
  } else if (label) {
    labelNode = label;
    if (typeof label === 'string') {
      labelNode = renderHighligherNode(label);
    }
  } else {
    labelNode = renderHighligherNode(String(value));
  }
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
    ref: ref,
    className: (0, _clsx.default)(className, cssScope(), selected && cssScope('-selected'), active && cssScope('-active'), disabled && cssScope('-disabled')),
    onClick: () => {
      toggle(value);
    },
    title: title
  }, restProps), /*#__PURE__*/_react.default.createElement(_Text.Text, {
    as: "div",
    className: cssScope('__content'),
    lineClamp: textOverflow ? 1 : undefined,
    sx: [{
      "props": {
        "Bf4jedk": _urmav4p.default
      }
    }]
  }, labelNode), /*#__PURE__*/_react.default.createElement("div", {
    className: cssScope('__check')
  }, selected ? menuItemSelectedIcon : null));
});

/**
 * For users
 */
const SelectOptionGroup = () => /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null);
exports.SelectOptionGroup = SelectOptionGroup;
Object.assign(SelectOptionGroup, {
  [_common.UD_SELECT_OPTION_GROUP_ID]: true,
  [_configProvider.COMPONENT_VERSION_ID]: "3.28.3"
});
const InternalSelectOptionGroup = exports.InternalSelectOptionGroup = /*#__PURE__*/(0, _react.forwardRef)((_ref2, ref) => {
  let {
    isFirst,
    label,
    children
  } = _ref2;
  const cssScope = (0, _common.useSelectListGroupCssScope)();
  return /*#__PURE__*/_react.default.createElement("div", {
    ref: ref,
    className: (0, _clsx.default)(cssScope(), isFirst && cssScope('-first'))
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(cssScope('__title'), (0, _stylex.default)(_stylex2.cssTypographyHeadline))
  }, label), children);
});
