// @universe-design/react/lib/select/Select.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Select = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _ListCheckBoldOutlined = _interopRequireDefault(require("@universe-design/icons-react/ListCheckBoldOutlined"));
var _react = _interopRequireDefault(require("react"));
var _reactUtilities = require("@universe-design/react-utilities");
var _util = require("../util");
var _common = require("./common");
var _selectGenerator = require("./selectGenerator");
var _SelectOverlay = require("./SelectOverlay");
const RefSelect = (0, _selectGenerator.selectGenerator)({
  selectorCssScope: 'select',
  dropdownCssScope: 'select',
  defaultOptionLabelProp: option => {
    var _ref, _option$labelRenderer, _option$labelRenderer2;
    return (_ref = (_option$labelRenderer = (_option$labelRenderer2 = option.labelRenderer) === null || _option$labelRenderer2 === void 0 ? void 0 : _option$labelRenderer2.call(option, '')) !== null && _option$labelRenderer !== void 0 ? _option$labelRenderer : option.label) !== null && _ref !== void 0 ? _ref : option.value;
  }
});
function selectChildren2Options(children) {
  return (0, _util.toFlatArray)(children).map((child, index) => {
    if ((0, _common.isSelectOption)(child)) {
      var _ref2;
      const {
        key,
        props: {
          children: _children,
          ...restProps
        }
      } = child;
      return {
        ...restProps,
        label: _children,
        key: (_ref2 = key !== null && key !== void 0 ? key : restProps.value) !== null && _ref2 !== void 0 ? _ref2 : `select-option-key-${index}`
      };
    } else if ((0, _common.isSelectOptionGroup)(child)) {
      const {
        key,
        props: {
          children: _children,
          ...restProps
        }
      } = child;
      return {
        ...restProps,
        key: key || `select-optgroup-key-${index}`,
        options: selectChildren2Options(_children)
      };
    }
    return null;
  }).filter(_util.isNonNullable);
}
const Select = exports.Select = /*#__PURE__*/_react.default.forwardRef((props, ref) => {
  const {
    options = selectChildren2Options(props.children),
    filterOption = true,
    defaultActiveFirstOption = true,
    menuItemSelectedIcon = /*#__PURE__*/_react.default.createElement(_ListCheckBoldOutlined.default, null),
    listHeight = 248,
    onPopupScroll,
    virtual = true,
    itemHeight = 32,
    filterSort,
    debounceTime,
    showSelectAll,
    onActiveOptionChange,
    ...baseSelectProps
  } = props;
  const internalOptions = _react.default.useMemo(() => options.reduce((pre, cur) => {
    if ('options' in cur) {
      for (const opt of cur.options) {
        pre.push(Object.assign({}, opt, {
          htmlTitle: opt.title
        }));
      }
    } else {
      pre.push(Object.assign({}, cur, {
        htmlTitle: cur.title
      }));
    }
    return pre;
  }, []), [options]);
  const forceUpdate = (0, _reactUtilities.useForceUpdate)();
  // We need force update here since popup dom is render async

  const refSelectRef = _react.default.useRef(null);
  const listRef = _react.default.useRef(null);
  _react.default.useImperativeHandle(ref, () => ({
    get input() {
      var _refSelectRef$current;
      return (_refSelectRef$current = refSelectRef.current) === null || _refSelectRef$current === void 0 ? void 0 : _refSelectRef$current.input;
    },
    focus() {
      var _refSelectRef$current2;
      return (_refSelectRef$current2 = refSelectRef.current) === null || _refSelectRef$current2 === void 0 ? void 0 : _refSelectRef$current2.focus();
    },
    blur() {
      var _refSelectRef$current3;
      return (_refSelectRef$current3 = refSelectRef.current) === null || _refSelectRef$current3 === void 0 ? void 0 : _refSelectRef$current3.blur();
    },
    // deprecated: use scrollInfoView instead
    scrollTo() {
      var _listRef$current;
      for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }
      return (_listRef$current = listRef.current) === null || _listRef$current === void 0 ? void 0 : _listRef$current.scrollIntoView(...args);
    },
    scrollIntoView() {
      var _listRef$current2;
      for (var _len2 = arguments.length, args = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
        args[_key2] = arguments[_key2];
      }
      return (_listRef$current2 = listRef.current) === null || _listRef$current2 === void 0 ? void 0 : _listRef$current2.scrollIntoView(...args);
    },
    setActiveOption(value) {
      var _listRef$current3;
      return (_listRef$current3 = listRef.current) === null || _listRef$current3 === void 0 ? void 0 : _listRef$current3.setActiveOption(value);
    }
  }), []);
  return /*#__PURE__*/_react.default.createElement(RefSelect, (0, _extends2.default)({}, baseSelectProps, {
    ref: refSelectRef,
    options: internalOptions,
    overlayNode: /*#__PURE__*/_react.default.createElement(_SelectOverlay.SelectOverlay, {
      ref: listRef,
      filterSort: filterSort,
      defaultActiveFirstOption: defaultActiveFirstOption,
      options: options,
      filterOption: filterOption,
      menuItemSelectedIcon: menuItemSelectedIcon,
      listHeight: listHeight,
      onPopupScroll: onPopupScroll,
      virtual: virtual,
      itemHeight: itemHeight,
      onMouseEnter: function () {
        forceUpdate();
      },
      onActiveOptionChange: onActiveOptionChange,
      debounceTime: debounceTime,
      showSelectAll: showSelectAll
    })
  }));
});
