// @universe-design/react/lib/select/Selector.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MultipleSelector = MultipleSelector;
exports.Selector = void 0;
exports.SingleSelector = SingleSelector;
var _usfh9mg = _interopRequireDefault(require("./../.stylex/usfh9mg"));
var _uxozh = _interopRequireDefault(require("./../.stylex/uxozh73"));
var _uh6bwfr = _interopRequireDefault(require("./../.stylex/uh6bwfr"));
var _u903f = _interopRequireDefault(require("./../.stylex/u903f20"));
var _uyhod = _interopRequireDefault(require("./../.stylex/uyhod33"));
var _clsx = _interopRequireDefault(require("clsx"));
var _react = _interopRequireWildcard(require("react"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _util = require("../util");
var _useSelection = require("./context/useSelection");
var _SelectContext = require("./context/SelectContext");
var _common = require("./common");
var _overflow = require("../overflow");
var _MultipleSelectorContext = require("./context/MultipleSelectorContext");
var _Input = _interopRequireDefault(require("../shared/components/Input"));
var _textOverflow = require("../shared/components/text-overflow");
var _hooks = require("../hooks");
var _FirstLineAlign = require("../shared/components/FirstLineAlign");
var _useTextOverflowPopperSidecar = require("../config-provider/hooks/useTextOverflowPopperSidecar");
var _stylex2 = require("../stylex");
var _dom = require("../util/dom");
var _text = require("../text");
//.uk73vx1{opacity:0;}
//.u5p0z4x{opacity:1;}
//.u1h5fybs{line-height:36px;}
//.u336tjw{line-height:28px;}
//.u8xlz6g{line-height:24px;}
/* eslint-disable react/destructuring-assignment */
const Input = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  const {
    id,
    inputRef,
    selectorCssScope: cssScope,
    inputValue,
    disabled,
    onInputChange,
    onInputCompositionStart,
    onInputCompositionEnd,
    onInputPaste,
    showSearch,
    tabIndex,
    inputting
  } = (0, _react.useContext)(_SelectContext.SelectContext);
  (0, _react.useImperativeHandle)(ref, () => inputRef.current);
  return /*#__PURE__*/_react.default.createElement(_Input.default, {
    id: id,
    tabIndex: tabIndex,
    role: "combobox",
    autoComplete: "off",
    disabled: disabled,
    type: "search",
    readOnly: !showSearch,
    ref: inputRef,
    className: (0, _clsx.default)(cssScope('__search__input'), inputting && cssScope('__search__input-inputting')),
    value: inputValue,
    onChange: onInputChange,
    onPaste: onInputPaste,
    onCompositionStart: onInputCompositionStart,
    onCompositionEnd: onInputCompositionEnd
  });
});
const AutoSizeInput = () => {
  const {
    selectorCssScope: cssScope,
    inputValue
  } = (0, _react.useContext)(_SelectContext.SelectContext);
  const measureRef = (0, _react.useRef)(null);
  const [inputWidth, setInputWidth] = (0, _react.useState)(0);
  (0, _hooks.useIsomorphicLayoutEffect)(() => {
    if (measureRef.current) {
      setInputWidth(measureRef.current.offsetWidth);
    }
  }, [inputValue]);
  return /*#__PURE__*/_react.default.createElement("span", {
    className: cssScope('__search__input_container'),
    style: {
      // TODO 不优雅
      width: inputWidth + 2
    }
  }, /*#__PURE__*/_react.default.createElement(Input, null), /*#__PURE__*/_react.default.createElement("span", {
    ref: measureRef,
    className: cssScope('__search__measure'),
    "aria-hidden": true
  }, inputValue, "\xA0"));
};
function MultipleSelector(_ref) {
  let {
    maxTagCount,
    maxTagRenderer
  } = _ref;
  const {
    valueSet,
    getOption,
    deselect,
    disabled: globalDisabled,
    selectorCssScope: cssScope,
    selectorLabelRenderer,
    tagRenderer
  } = _react.default.useContext(_SelectContext.SelectContext);
  const autoSizeInputNode = (0, _react.useMemo)(() => /*#__PURE__*/_react.default.createElement(AutoSizeInput, null), []);
  return /*#__PURE__*/_react.default.createElement(_overflow.Overflow, {
    className: (0, _clsx.default)(cssScope('__search')),
    maxCount: maxTagCount,
    suffixClassName: cssScope('__search__suffix'),
    suffixWrap: false,
    keys: [...valueSet].map(_common.getReactKey),
    itemRenderer: opt => {
      const _opt = opt;
      const option = getOption(_opt);
      const disabled = (0, _useSelection.isDisabledOption)(option, globalDisabled);
      return tagRenderer({
        title: option.htmlTitle,
        value: option.value,
        disabled,
        closable: !disabled,
        onClose: () => {
          deselect(_opt);
        },
        children: selectorLabelRenderer(_opt)
      });
    },
    restRenderer: omittedKeys => maxTagRenderer({
      disabled: globalDisabled
    }, omittedKeys.map(opt => {
      const _opt = opt;
      const option = getOption(_opt);
      const disabled = (0, _useSelection.isDisabledOption)(option, globalDisabled);
      return {
        value: option.value,
        disabled,
        closable: !disabled,
        onClose: () => {
          deselect(_opt);
        },
        children: selectorLabelRenderer(_opt)
      };
    })),
    suffixRenderer: () => autoSizeInputNode
  });
}
function SingleSelector() {
  const {
    selectorCssScope: cssScope
  } = _react.default.useContext(_SelectContext.SelectContext);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: cssScope('__search')
  }, /*#__PURE__*/_react.default.createElement(Input, null));
}
function SelectorForwardRefRenderFunction(_ref2, ref) {
  let {
    size,
    placeholder,
    bordered,
    error,
    maxTagCount,
    allowClear,
    showArrow,
    maxTagRenderer,
    clearIcon,
    suffixIcon
  } = _ref2;
  const {
    multiple,
    focusState,
    overlayVisible,
    clear,
    valueSet,
    inputValue,
    disabled,
    selectorCssScope: cssScope,
    selectorLabelRenderer,
    setOverlayVisible,
    inputRef,
    inputting,
    showSearch
  } = _react.default.useContext(_SelectContext.SelectContext);
  const {
    maxHeight: multipleSelectorMaxHeight
  } = (0, _MultipleSelectorContext.useMultipleSelectorContext)();
  const {
    isMouseOver,
    handleMouseEnter,
    handleMouseLeave
  } = (0, _hooks.useMouseOver)();
  const TextOverflowPopper = (0, _useTextOverflowPopperSidecar.useTextOverflowPopperSidecar)();
  const placeholderRef = _react.default.useRef(null);
  const [placeholderTooltipVisible, setPlaceholderTooltipVisible] = _react.default.useState(false);
  let selectorNode;
  if (multiple) {
    selectorNode = /*#__PURE__*/_react.default.createElement(MultipleSelector, {
      maxTagCount: maxTagCount,
      maxTagRenderer: maxTagRenderer
    });
  } else {
    selectorNode = /*#__PURE__*/_react.default.createElement(SingleSelector, null);
  }
  const isSelectValueEmpty = !valueSet.size;
  const textVariant = {
    small: 'body2',
    medium: 'body0',
    large: 'title5'
  }[size];
  // 是否应该显示 placeholder
  const isPlaceholderVisible = isSelectValueEmpty && Boolean(placeholder) && !inputValue;
  /**
   * 这儿加个 key
   * 在切换 placeholder 和 selectItem 时, react 有可能会复用节点
   * 有可能切换到 selectItem 的一瞬间, 由于还是复用的 placeholder 的节点
   * 所以那一瞬间继承了 placeholder 的 class, 导致文字颜色有一瞬间是 placeholder 的颜色
   * 此时如果这个节点上有个 transition 的过渡, 就会导致该节点的颜色闪一下
   */
  let placeholderNode = /*#__PURE__*/_react.default.createElement(_text.Text, {
    key: "placeholder",
    as: "div",
    variant: textVariant,
    className: (0, _clsx.default)(cssScope('__placeholder'), (0, _stylex.default)(_stylex2.sxOverflowEllipsis)),
    ref: placeholderRef
  }, placeholder);
  if (inputValue) {
    if (multiple && !isSelectValueEmpty) {
      placeholderNode = null;
    } else {
      placeholderNode = /*#__PURE__*/_react.default.createElement("span", {
        className: cssScope('__placeholder')
      });
    }
  } else if (!isSelectValueEmpty) {
    if (multiple) {
      placeholderNode = null;
    } else {
      // first item
      const selectItem = valueSet.keys().next().value;
      placeholderNode = /*#__PURE__*/_react.default.createElement(_textOverflow.TextOverflow, {
        key: "selectItem",
        as: "div",
        className: (0, _clsx.default)(cssScope('__selectItem'), inputting && cssScope('__selectItem-inputting')),
        lineClamp: 1,
        variant: textVariant,
        disabled: disabled
      }, selectorLabelRenderer(selectItem));
    }
  }
  const showClear = !isSelectValueEmpty && !disabled && allowClear;

  /**
   * 兼容其他 DS
   */

  return /*#__PURE__*/_react.default.createElement("div", {
    ref: ref,
    onMouseEnter: handleMouseEnter,
    onMouseLeave: handleMouseLeave,
    className: (0, _clsx.default)(cssScope(), cssScope(`-${(0, _util.getSizeAbbr)(size)}`), focusState && cssScope('-focus'), overlayVisible && cssScope('-open'), !showSearch && cssScope('-readOnly'), cssScope(`-border-${function () {
      if (bordered === true) {
        return 'normal';
      } else if (bordered === false) {
        return 'noborder';
      }
      return bordered;
    }()}`), !isSelectValueEmpty && cssScope('-not-empty'), showClear && cssScope('-allowClear'),
    // isMouseOver && showClear && cssScope('-showClear'),
    multiple && cssScope('-multiple'), disabled && cssScope('-disabled'), error && cssScope('-error'), !showArrow && cssScope('-hide-arrow')),
    onClick: _ref3 => {
      let {
        target
      } = _ref3;
      // 这里要用 onClick 不能用 mousedown 和 mouseup，因为 multiple 模式下 tag 的 onClick 做了 stopPropagation
      if (target !== inputRef.current) {
        var _inputRef$current;
        (_inputRef$current = inputRef.current) === null || _inputRef$current === void 0 ? void 0 : _inputRef$current.focus();
      }
      setOverlayVisible();
    },
    style: {
      maxHeight: multiple ? multipleSelectorMaxHeight : undefined
    }
  }, /*#__PURE__*/_react.default.createElement(_FirstLineAlign.EmptyElement, {
    sx: [{
      small: {
        "props": {
          "Bg96gwp": _uyhod.default
        }
      },
      medium: {
        "props": {
          "Bg96gwp": _u903f.default
        }
      },
      large: {
        "props": {
          "Bg96gwp": _uh6bwfr.default
        }
      }
    }[size]]
  }), /*#__PURE__*/_react.default.createElement(TextOverflowPopper, {
    title: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, placeholder),
    checkOverflow: () => placeholderRef.current ? (0, _dom.isOverflowed)(placeholderRef.current) : false,
    visible: isPlaceholderVisible && placeholderTooltipVisible,
    onVisibleChange: visible => {
      setPlaceholderTooltipVisible(visible);
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: cssScope('__content'),
    onMouseLeave: () => {
      setPlaceholderTooltipVisible(false);
    }
  }, placeholderNode, selectorNode)), showClear ? /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(cssScope('__clear'), (0, _stylex.default)(isMouseOver ? {
      "props": {
        "abs64n": _uxozh.default
      }
    } : {
      "props": {
        "abs64n": _usfh9mg.default
      }
    })),
    onMouseDown: e => {
      e.stopPropagation();
      clear();
    },
    onClick: e => {
      // 防止点击冒泡触发 overlay 打开
      e.stopPropagation();
    }
  }, clearIcon) : null, showArrow ? /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(cssScope('__arrow'))
  }, suffixIcon) : null);
}
// TODO typing is not safe!
const Selector = exports.Selector = /*#__PURE__*/_react.default.forwardRef(SelectorForwardRefRenderFunction);
