// @universe-design/react/lib/select/SelectAsync.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SelectAsync = void 0;
exports.useSelectAsync = useSelectAsync;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _rxjs = require("rxjs");
var _operators = require("rxjs/operators");
var _Select = require("./Select");
var _reactUtilities = require("@universe-design/react-utilities");
var _hooks = require("../hooks");
var _useLoading = require("../loading/useLoading");
var _util = require("../util");
var Actions = /*#__PURE__*/function (Actions) {
  Actions[Actions["StartFetching"] = 0] = "StartFetching";
  Actions[Actions["GotOptionsSuccessfully"] = 1] = "GotOptionsSuccessfully";
  Actions[Actions["GotOptionsFailed"] = 2] = "GotOptionsFailed";
  return Actions;
}(Actions || {});
function useSelectAsync(props) {
  const {
    defaultOptions = [],
    loadOptions,
    onSearch,
    debounceTime = 800,
    loadingDelay = 200,
    onBeforeLoadOptions,
    onLoadOptionsError,
    onLoadOptionsSuccess,
    // omit
    options: optionsProp,
    filterOption: filterOptionProp,
    loading: loadingProp,
    ...restProps
  } = props;
  process.env.NODE_ENV !== "production" ? (0, _util.warning)(optionsProp === undefined, (0, _util.devWarning)('useSelectAsync', 'options will be ignored in async select')) : void 0;
  process.env.NODE_ENV !== "production" ? (0, _util.warning)(loadingProp === undefined, (0, _util.devWarning)('useSelectAsync', 'loading will be ignored in async select')) : void 0;
  process.env.NODE_ENV !== "production" ? (0, _util.warning)(filterOptionProp === undefined, (0, _util.devWarning)('useSelectAsync', 'filterOption will be ignored in async select')) : void 0;
  const initialState = (0, _reactUtilities.useConstant)(() => {
    var _restProps$searchValu;
    return {
      loading: false,
      options: Array.isArray(defaultOptions) ? defaultOptions : [],
      searchValue: (_restProps$searchValu = restProps.searchValue) !== null && _restProps$searchValu !== void 0 ? _restProps$searchValu : ''
    };
  });
  const [state, dispatch] = (0, _react.useReducer)((_state, action) => {
    switch (action.type) {
      case Actions.StartFetching:
        (0, _util.isFunction)(onBeforeLoadOptions) && onBeforeLoadOptions();
        return {
          ..._state,
          loading: true
        };
      case Actions.GotOptionsSuccessfully:
        (0, _util.isFunction)(onLoadOptionsSuccess) && onLoadOptionsSuccess(action.payload);
        return {
          ..._state,
          options: action.payload,
          loading: false
        };
      case Actions.GotOptionsFailed:
        (0, _util.isFunction)(onLoadOptionsError) && onLoadOptionsError(action.payload);
        return {
          ..._state,
          loading: false,
          error: action.payload
        };
      default:
        throw new Error('Action type not found');
    }
  }, initialState);
  const [delayedLoading] = (0, _useLoading.useLoading)({
    loading: state.loading,
    delay: loadingDelay
  });
  const isMounted = (0, _hooks.useMountedState)();
  // const isFirstMount = useFirstMountState();

  // Observable
  const event$ = (0, _reactUtilities.useConstant)(() => new _rxjs.Subject());
  const state$ = (0, _reactUtilities.useConstant)(() => new _rxjs.BehaviorSubject(initialState));
  (0, _react.useEffect)(() => {
    const value$ = event$.pipe((0, _operators.debounce)(ev => ev.delay ? (0, _rxjs.timer)(ev.delay) : (0, _rxjs.of)(ev)), (0, _operators.withLatestFrom)(state$), (0, _operators.tap)(_ref => {
      let [ev] = _ref;
      if (isMounted() && ev.showLoading) {
        dispatch({
          type: Actions.StartFetching
        });
      }
    }), (0, _operators.switchMap)(_ref2 => {
      let [ev, _state] = _ref2;
      if (!(0, _util.isFunction)(loadOptions)) {
        return (0, _rxjs.of)(_state);
      }
      try {
        const ret = loadOptions(ev.searchValue, _state.options);
        const mapOptions = (0, _operators.map)(_options => ({
          ..._state,
          options: _options
        }));
        // eslint-disable-next-line no-nested-ternary
        return ((0, _rxjs.isObservable)(ret) ? ret : (0, _util.isPromise)(ret) ? (0, _rxjs.from)(ret) : (0, _rxjs.of)(ret)).pipe(mapOptions, (0, _operators.catchError)(err => {
          console.error('Unexpected error occured when loading options: ', err, 'Please handle it manually');
          if (isMounted()) {
            dispatch({
              type: Actions.GotOptionsFailed,
              payload: err
            });
          }
          return (0, _rxjs.of)(_state);
        }));
      } catch (err) {
        return (0, _rxjs.throwError)(err);
      }
    }));
    const subscription = value$.subscribe(value => {
      if (isMounted()) {
        state$.next(value);
        dispatch({
          type: Actions.GotOptionsSuccessfully,
          payload: value.options
        });
      }
    }, err => {
      console.log('value$ complete with error: ', err);
    });
    return () => {
      subscription.unsubscribe();
    };
  }, [event$, isMounted, loadOptions, state$]); // immutable forever

  (0, _hooks.useEffectOnce)(() => {
    if (defaultOptions === false || Array.isArray(defaultOptions)) {
      return;
    }
    if ((0, _util.isFunction)(loadOptions)) {
      event$.next({
        searchValue: initialState.searchValue,
        delay: 0,
        showLoading: false
      });
    }
    return () => {
      state$.complete();
      event$.complete();
    };
  });
  const handleSearch = (0, _react.useCallback)((v, meta) => {
    event$.next({
      searchValue: v,
      delay: debounceTime,
      showLoading: true
    });
    (0, _util.isFunction)(onSearch) && onSearch(v, meta);
  }, [debounceTime, event$, onSearch]);
  const additionalProps = {
    loading: delayedLoading,
    options: state.options,
    onSearch: handleSearch,
    filterOption: false
  };
  return {
    ...additionalProps,
    ...restProps
  };
}
const SelectAsync = exports.SelectAsync = /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
  const asyncProps = useSelectAsync(props);
  // FIXME error TS2769: No overload matches this call.
  return /*#__PURE__*/_react.default.createElement(_Select.Select, (0, _extends2.default)({
    ref: ref
  }, asyncProps));
});
