// @universe-design/react/es/overflow/index.js
import clsx from 'clsx';
import React from 'react';
import { useCssScope } from '../config-provider/hooks';
import { useOverflowItem, OverflowContext, useOverflow, useOverflowRest, useOverflowSuffix } from './hooks';
import './style';
const _OverflowItem = _ref => {
  let {
    value,
    index,
    children,
    getItemStyle
  } = _ref;
  const cssScope = useCssScope('overflow__item');
  const {
    ref,
    isVisible
  } = useOverflowItem({
    key: value,
    index
  });
  const itemStyle = getItemStyle === null || getItemStyle === void 0 ? void 0 : getItemStyle(value);
  const order = isVisible ? index : Number.MAX_SAFE_INTEGER;
  return /*#__PURE__*/React.createElement("div", {
    className: cssScope(),
    style: {
      order,
      height: isVisible ? undefined : 0,
      overflowY: isVisible ? undefined : 'hidden',
      pointerEvents: isVisible ? undefined : 'none',
      ...itemStyle
    }
  }, /*#__PURE__*/React.createElement("div", {
    ref: ref
  }, children));
};
if (process.env.NODE_ENV !== "production") {
  _OverflowItem.displayName = 'OverflowItem';
}
const OverflowItem = /*#__PURE__*/React.memo(_OverflowItem);
const OverflowRest = /*#__PURE__*/React.memo(_ref2 => {
  let {
    children,
    restStyle
  } = _ref2;
  const {
    ref,
    countBeforeRest,
    isVisible,
    isResponsive,
    restPosition
  } = useOverflowRest();
  const cssScope = useCssScope('overflow__rest');
  const order = isVisible ? countBeforeRest - 1 : Number.MAX_SAFE_INTEGER;
  return /*#__PURE__*/React.createElement("div", {
    ref: ref,
    className: cssScope(''),
    style: {
      // 这里是为了修复在父级 fit-content 时，rest 节点没有展示也会占空间的问题
      position: isVisible ? undefined : 'absolute',
      order: isResponsive ? order : restPosition - 1,
      height: isVisible ? undefined : 0,
      overflowY: isVisible ? undefined : 'hidden',
      pointerEvents: isVisible ? undefined : 'none',
      ...restStyle
    }
  }, children);
});
const OverflowSuffix = /*#__PURE__*/React.memo(_ref3 => {
  let {
    className,
    children
  } = _ref3;
  const {
    ref,
    suffixFixedStart,
    countAfterRest,
    countBeforeRest,
    isResponsive,
    restWidth,
    suffixWrap
  } = useOverflowSuffix();
  const cssScope = useCssScope('overflow__suffix');
  return /*#__PURE__*/React.createElement("div", {
    className: clsx(className, cssScope('')),
    ref: ref,
    style: {
      order: isResponsive ? countBeforeRest + countAfterRest - 1 : Number.MAX_SAFE_INTEGER,
      ...(suffixFixedStart !== null && isResponsive ? {
        position: 'absolute',
        left: suffixFixedStart
      } : {}),
      maxWidth: suffixWrap ? undefined : `calc(100% - ${restWidth}px)` // 限制 suffix 不换行时的最大宽度
    }
  }, children);
});
const defaultItemRenderer = key => /*#__PURE__*/React.createElement("div", {
  style: {
    border: '1px red solid'
  }
}, Array(Number(key) + 1).fill(0).map((_, i) => i).join(','));
const defaultRestRenderer = omittedKeys => /*#__PURE__*/React.createElement("div", {
  style: {
    border: '1px blue solid'
  }
}, `+${omittedKeys.length}`);
export const Overflow = _ref4 => {
  let {
    maxCount,
    keys,
    style,
    getItemStyle,
    restStyle,
    className,
    itemRenderer = defaultItemRenderer,
    restRenderer = defaultRestRenderer,
    suffixRenderer,
    suffixClassName,
    restPosition,
    addonRenderer,
    alwaysObserveResize = false,
    suffixWrap = true
  } = _ref4;
  const cssScope = useCssScope('overflow');
  const {
    ref,
    dispatch,
    isResponsive,
    ...state
  } = useOverflow({
    maxCount,
    keys,
    hasSuffix: Boolean(suffixRenderer),
    restPosition,
    suffixWrap
  });
  const showRest = isResponsive || typeof maxCount === 'number' && keys.length > maxCount;
  return /*#__PURE__*/React.createElement(OverflowContext.Provider, {
    value: {
      dispatch,
      isResponsive,
      ...state,
      alwaysObserveResize
    }
  }, /*#__PURE__*/React.createElement("div", {
    ref: ref,
    className: clsx(cssScope(), className),
    style: style
  }, [...state.mergedKeys.beforeRest, ...state.mergedKeys.afterRest].map((key, i) => /*#__PURE__*/React.createElement(OverflowItem, {
    key: key,
    value: key,
    index: i,
    getItemStyle: getItemStyle
  }, itemRenderer(key))), showRest && /*#__PURE__*/React.createElement(OverflowRest, {
    restStyle: restStyle
  }, restRenderer(state.omittedKeys)), suffixRenderer && /*#__PURE__*/React.createElement(OverflowSuffix, {
    className: suffixClassName
  }, suffixRenderer({
    omittedKeys: state.omittedKeys
  })), addonRenderer === null || addonRenderer === void 0 ? void 0 : addonRenderer({
    omittedKeys: state.omittedKeys
  })));
};
