// @universe-design/react/es/date-picker/panels/base/Body.js
import _extends from "@babel/runtime/helpers/esm/extends";
import clsx from 'clsx';
import * as React from 'react';
import { useCssScope } from '../../../config-provider/hooks';
import Cell from '../../cell';
import { useDateCalendarUIContext } from '../DateCalendar/DateCalendarUIContext';
function Body(props) {
  const {
    className,
    cellClassName,
    rowNum,
    colNum,
    baseDate,
    autoHeight,
    getCellDate = date => date,
    dateRender = () => null,
    getCellStatus,
    // headerCells,
    children
  } = props;
  const prefixCls = useCssScope('picker')();
  const bodyPrefixCls = `${prefixCls}-panel-body`;
  const {
    onDateMouseEnter,
    onDateMouseLeave,
    onSelect,
    onContentMouseEnter,
    onContentMouseLeave
  } = useDateCalendarUIContext();
  const rows = [];
  for (let rowIndex = 0; rowIndex < rowNum; rowIndex++) {
    const isRowSelected = [...new Array(colNum)].some((_, colIndex) => getCellStatus(getCellDate(baseDate, rowIndex * colNum + colIndex)).isSelected);
    rows.push(/*#__PURE__*/React.createElement("div", {
      className: clsx(`${bodyPrefixCls}-row`, isRowSelected && `${bodyPrefixCls}-row--selected`),
      key: rowIndex
    }, [...new Array(colNum)].map((_, colIndex) => {
      const offset = rowIndex * colNum + colIndex;
      const cellDate = getCellDate(baseDate, offset);
      const cellStatus = getCellStatus(cellDate);
      const {
        isDisabled
      } = cellStatus;
      const preCellDate = getCellDate(baseDate, offset - 1);
      const nextCellDate = getCellDate(baseDate, offset + 1);
      const prevCellStatus = getCellStatus(preCellDate);
      const nextCellStatus = getCellStatus(nextCellDate);
      return /*#__PURE__*/React.createElement(React.Fragment, {
        key: colIndex
      }, /*#__PURE__*/React.createElement(Cell, _extends({
        onClick: () => {
          if (!isDisabled) {
            onSelect(cellDate);
          }
        },
        onMouseEnter: () => {
          if (!isDisabled) {
            onDateMouseEnter(cellDate);
          }
        },
        onMouseLeave: () => {
          if (!isDisabled) {
            onDateMouseLeave(cellDate);
          }
        },
        prevCellStatus: prevCellStatus,
        nextCellStatus: nextCellStatus,
        className: cellClassName,
        colIndex: colIndex,
        colNum: colNum
      }, cellStatus), dateRender(cellDate)));
    })));
  }
  return /*#__PURE__*/React.createElement("div", {
    className: clsx(bodyPrefixCls, autoHeight && `${prefixCls}-auto-height`, className)
  }, children, /*#__PURE__*/React.createElement("div", {
    className: clsx(`${bodyPrefixCls}-content`, 'notranslate'),
    onMouseEnter: onContentMouseEnter,
    onMouseLeave: onContentMouseLeave
  }, rows));
}
export default Body;
