// @universe-design/react/es/date-picker/panels/base/utils.js
export const YEAR_COUNT = 20;
export const getYearStart = date => {
  const newDate = date.clone();
  return newDate.year(Math.floor(newDate.year() / YEAR_COUNT) * YEAR_COUNT);
};
export const getYearRange = date => {
  const start = getYearStart(date).year();
  return [start, start + YEAR_COUNT - 1].join(' - ');
};
function generatorDateByMode(date, mode, offset) {
  const newDate = date.clone();
  switch (mode) {
    case 'date':
      return newDate.add(offset, 'month');
    case 'month':
      return newDate.add(offset, 'year');
    case 'year':
      // return newDate.add(offset * YEAR_COUNT, 'year');
      return getYearStart(newDate).add(offset * YEAR_COUNT, 'year');
    default:
      return newDate;
  }
}
export const getNext = (date, mode) => generatorDateByMode(date, mode, 1);
export const getPrev = (date, mode) => generatorDateByMode(date, mode, -1);
export const getHigherLevelMode = mode => {
  const levels = ['date', 'month', 'year'];
  const index = levels.findIndex(m => m === mode);
  if (index < levels.length - 1) {
    return levels[index + 1];
  }
  return null;
};
export function isInRange(date, range, mode) {
  return date.isSameOrAfter(range[0], mode) && date.isSameOrBefore(range[1], mode);
}
