// @universe-design/react/src/skeleton/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Basic from '../demo/0.basic';
import C1SkeletonItem from '../demo/1.skeletonItem';
import C2Table from '../demo/2.table';
import C3Animation from '../demo/3.animation';
if (!(C0Basic as Story).storyName) {
  (C0Basic as Story).storyName = '0. 基础用法';
  (C0Basic as Story).parameters = {
    docs: {
      description: {
        story: '简单的用法。\n\n',
      },
    },
  };
}
if (!(C1SkeletonItem as Story).storyName) {
  (C1SkeletonItem as Story).storyName = '1. 占位元素';
  (C1SkeletonItem as Story).parameters = {
    docs: {
      description: {
        story:
          '提供 Rectangle、Circle、Paragraph 三种占位元素，提供基础的默认样式，在 placeholder 中自由修改样式，自由布局。\n\n- 矩形块默认样式：width: `100%`，height: `14px`, marginBottom: `16px`\n- 圆形块默认样式：width: `32px`，height: `32px`\n\n',
      },
    },
  };
}
if (!(C2Table as Story).storyName) {
  (C2Table as Story).storyName = '2. 在表格中使用';
  (C2Table as Story).parameters = {
    docs: {
      description: {
        story: '通过表格 Column 中的 `title` 和 `render` 属性实现表格骨架屏。\n\n',
      },
    },
  };
}
if (!(C3Animation as Story).storyName) {
  (C3Animation as Story).storyName = '3. 加载动画';
  (C3Animation as Story).parameters = {
    docs: {
      description: {
        story: '通过 `animation` 属性开启或关闭动画。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Skeleton/官网示例',
  tags: ['autodocs'],
};

export { C0Basic, C1SkeletonItem, C2Table, C3Animation };
