// @universe-design/react/es/carousel/CarouselSwitcher.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.ucq9r29{right:24px;}
import u1aqot9l from "./../.stylex/u1aqot9l";
//.u1euv43f{position:absolute;}
import u1u1p7mj from "./../.stylex/u1u1p7mj";
//.u1ycc5zk{left:24px;}
import u1utdkg1 from "./../.stylex/u1utdkg1";
//.u1i1t8d1{top:50%;}
import ufibavv from "./../.stylex/ufibavv";
//.u188r07x{transform:translateY(-50%);}
import ulanhzm from "./../.stylex/ulanhzm";
//.ueqmc2u{width:40px;}
import uwf09ar from "./../.stylex/uwf09ar";
//.ubhnoac{height:40px;}
import ubr2xhg from "./../.stylex/ubr2xhg";
//.u4ybsrx{font-size:16px;}
import u1o367ev from "./../.stylex/u1o367ev";
//.u1w9dchk{width:28px;}
import uggvdc4 from "./../.stylex/uggvdc4";
//.uxldao9{height:28px;}
import un6nsv3 from "./../.stylex/un6nsv3";
//.uses1vf{font-size:14px;}
import u19bueuo from "./../.stylex/u19bueuo";
//.u22iagw{display:flex;}
import ukchsq7 from "./../.stylex/ukchsq7";
//.u122n59{align-items:center;}
import ubyhozq from "./../.stylex/ubyhozq";
//.u4d9j23{justify-content:center;}
import ut7q8es from "./../.stylex/ut7q8es";
//.u19ibtj8{color:@static-white;}
import uathu2q from "./../.stylex/uathu2q";
//.u1c04qm4{background-color:@static-black-50;}
import u5f56vu from "./../.stylex/u5f56vu";
//.u1ern45e{border-top-style:none;}
import ugowvl5 from "./../.stylex/ugowvl5";
//.u1n71otn{border-right-style:none;}
import u1h28k4f from "./../.stylex/u1h28k4f";
//.u1h8hb77{border-bottom-style:none;}
import u1qg4c09 from "./../.stylex/u1qg4c09";
//.u1deefiw{border-left-style:none;}
import ugmgj40 from "./../.stylex/ugmgj40";
//.u1k6fduh{cursor:pointer;}
import u12o6wmt from "./../.stylex/u12o6wmt";
//.u1mf1pf4{transition-property:background-color,color;}
import u5kb0jy from "./../.stylex/u5kb0jy";
//.ubf55e9{transition-duration:0.1s,0.1s;}
import up10246 from "./../.stylex/up10246";
//.u6mlncw:hover{background-color:@static-black-80;}
import uit90k4 from "./../.stylex/uit90k4";
//.uxdupyk:active{background-color:@static-black-80;}
import u5nolws from "./../.stylex/u5nolws";
//.u2hkw1w:focus-visible{outline-style:none;}
import u1ul1s2h from "./../.stylex/u1ul1s2h";
//.u17ot7lm:focus-visible{background-color:@static-black-80;}
import u15mrnzb from "./../.stylex/u15mrnzb";
import LeftOutlined from '@universe-design/icons-react/LeftOutlined';
import RightOutlined from '@universe-design/icons-react/RightOutlined';
import stylex from '@universe-design/stylex';
import clsx from 'clsx';
import * as React from 'react';
import { sxBorderRadius, sxTransitionTimingFunction } from '../stylex';
import { forwardHTMLRef } from '../util/forwardRef';
import { useCarouselContext } from './CarouselContext';
import { useCarouselManagerContext } from './CarouselManager';
/**
 * TODO: 等 button 迁移 stylex 之后用 ud button 改写
 */
const CarouselSwitcher = forwardHTMLRef((_ref, ref) => {
  let {
    // omit
    children,
    // props
    className,
    sx,
    icon,
    size = 'medium',
    ...buttonProps
  } = _ref;
  return /*#__PURE__*/React.createElement("button", _extends({
    ref: ref,
    type: "button",
    className: clsx(className, stylex({
      "props": {
        "mc9l5x": ukchsq7,
        "Bt984gj": ubyhozq,
        "Brf1p80": ut7q8es,
        "sj55zd": uathu2q,
        "De3pzq": u5f56vu,
        "icvyot": ugowvl5,
        "vrafjx": u1h28k4f,
        "oivjwe": u1qg4c09,
        "wvpqe5": ugmgj40,
        "Bceei9c": u12o6wmt,
        "Bmy1vo4": u5kb0jy,
        "B3o57yi": up10246,
        "Jwef8y": uit90k4,
        "ecr2s2": u5nolws,
        "B486eqv": u1ul1s2h,
        "xyvl0c": u15mrnzb
      }
    }, sxBorderRadius.xl, sxTransitionTimingFunction.standard, {
      small: {
        "props": {
          "a9b677": uggvdc4,
          "Bqenvij": un6nsv3,
          "Be2twd7": u19bueuo
        }
      },
      medium: {
        "props": {
          "a9b677": uwf09ar,
          "Bqenvij": ubr2xhg,
          "Be2twd7": u1o367ev
        }
      }
    }[size], sx))
  }, buttonProps), icon);
});

/**
 * @public
 * @docgen
 */

export const CarouselPrevSwitcher = forwardHTMLRef((_ref2, ref) => {
  let {
    // omit
    children,
    // props
    onClick,
    sx,
    ...buttonProps
  } = _ref2;
  const {
    rootPrefixCls
  } = useCarouselContext();
  const manager = useCarouselManagerContext();
  return /*#__PURE__*/React.createElement(CarouselSwitcher, _extends({
    ref: ref,
    icon: /*#__PURE__*/React.createElement(LeftOutlined, null),
    className: `${rootPrefixCls}__prev-switcher`
  }, buttonProps, {
    sx: [{
      "props": {
        "qhf8xq": u1u1p7mj,
        "oyh7mz": u1utdkg1,
        "Bhzewxz": ufibavv,
        "Bz10aip": ulanhzm
      }
    }, sx],
    onClick: ev => {
      onClick === null || onClick === void 0 ? void 0 : onClick(ev);
      if (!manager.canUserInteract) {
        return;
      }
      manager.setSlideIndexByStep(-1);
    }
  }));
});
export const CarouselNextSwitcher = forwardHTMLRef((_ref3, ref) => {
  let {
    // omit
    children,
    // props
    onClick,
    sx,
    ...buttonProps
  } = _ref3;
  const {
    rootPrefixCls
  } = useCarouselContext();
  const manager = useCarouselManagerContext();
  return /*#__PURE__*/React.createElement(CarouselSwitcher, _extends({
    ref: ref,
    icon: /*#__PURE__*/React.createElement(RightOutlined, null),
    className: `${rootPrefixCls}__next-switcher`
  }, buttonProps, {
    sx: [{
      "props": {
        "qhf8xq": u1u1p7mj,
        "j35jbq": u1aqot9l,
        "Bhzewxz": ufibavv,
        "Bz10aip": ulanhzm
      }
    }],
    onClick: ev => {
      onClick === null || onClick === void 0 ? void 0 : onClick(ev);
      if (!manager.canUserInteract) {
        return;
      }
      manager.setSlideIndexByStep(1);
    }
  }));
});
