// @universe-design/react/es/carousel/Carousel.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.u1ktbn1t{left:50%;}
import u1jmni06 from "./../.stylex/u1jmni06";
//.u13gfj74{transform:translateX(-50%);}
import u9yw1hi from "./../.stylex/u9yw1hi";
//.u1euv43f{position:absolute;}
import u1u1p7mj from "./../.stylex/u1u1p7mj";
//.uf2sm71{height:48px;}
import us2xlyu from "./../.stylex/us2xlyu";
//.u1xynx9j{left:0px;}
import u184z4yl from "./../.stylex/u184z4yl";
//.u10k790i{right:0px;}
import u5j3c7v from "./../.stylex/u5j3c7v";
//.u5gq2j6{bottom:0px;}
import u8rtfz1 from "./../.stylex/u8rtfz1";
//.ufsbof8{background-image:linear-gradient(0deg, @static-black-15 0%, transparent 80.21%);}
import u1j6p86f from "./../.stylex/u1j6p86f";
//.u22iagw{display:flex;}
import ukchsq7 from "./../.stylex/ukchsq7";
//.u1vx9l62{flex-direction:column;}
import uohwdt from "./../.stylex/uohwdt";
//.u1l02sjl{height:100%;}
import udg2fzc from "./../.stylex/udg2fzc";
import * as React from 'react';
import stylex from '@universe-design/stylex';
import clsx from 'clsx';
import { CarouselContextProvider, useCarouselContext, useCarousel } from './CarouselContext';
import { CarouselWindow } from './CarouselWindow';
import { CarouselPrevSwitcher, CarouselNextSwitcher } from './CarouselSwitcher';
import { CarouselIndicator } from './CarouselIndicator';
import { CarouselManagerContextProvider } from './CarouselManager';
import { carouselDefaults } from './defaults';
import { FadeCarouselTransitionDescriptor } from './transition-descriptors/fade';
import { PushCarouselTransitionDescriptor } from './transition-descriptors/push';
const ComposableCarousel = _ref => {
  let {
    children,
    className,
    style,
    ...useCarouselProps
  } = _ref;
  const {
    manager,
    contextValue
  } = useCarousel(useCarouselProps);
  const {
    rootPrefixCls
  } = contextValue;
  return /*#__PURE__*/React.createElement(CarouselManagerContextProvider, {
    value: manager
  }, /*#__PURE__*/React.createElement(CarouselContextProvider, {
    value: contextValue
  }, /*#__PURE__*/React.createElement("div", {
    className: clsx(rootPrefixCls, stylex({
      "props": {
        "mc9l5x": ukchsq7,
        "Beiy3e4": uohwdt,
        "Bqenvij": udg2fzc
      }
    }), className),
    style
  }, children)));
};
const CarouselMask = () => {
  const {
    rootPrefixCls
  } = useCarouselContext();
  return /*#__PURE__*/React.createElement("div", {
    className: clsx(`${rootPrefixCls}__mask`, stylex({
      "props": {
        "qhf8xq": u1u1p7mj,
        "Bqenvij": us2xlyu,
        "oyh7mz": u184z4yl,
        "j35jbq": u5j3c7v,
        "B5kzvoi": u8rtfz1,
        "Bcmaq0h": u1j6p86f
      }
    }))
  });
};

/**
 * @public
 * @docgen
 */

/**
 * @public
 */
export const Carousel = _ref2 => {
  let {
    children,
    indicatorProps,
    indicatorPosition = carouselDefaults.indicatorPosition,
    switcherProps,
    transition = carouselDefaults.transition,
    ...sharedProps
  } = _ref2;
  return /*#__PURE__*/React.createElement(ComposableCarousel, _extends({}, sharedProps, {
    slides: children,
    transitionDescriptor: (() => {
      if (transition === 'fade') {
        return new FadeCarouselTransitionDescriptor();
      }
      return new PushCarouselTransitionDescriptor();
    })()
  }), /*#__PURE__*/React.createElement(CarouselWindow, null, /*#__PURE__*/React.createElement(CarouselPrevSwitcher, switcherProps), /*#__PURE__*/React.createElement(CarouselNextSwitcher, switcherProps), indicatorPosition === 'inner-bottom' && /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(CarouselMask, null), /*#__PURE__*/React.createElement(CarouselIndicator, _extends({
    theme: "light"
  }, indicatorProps, {
    sx: [{
      "props": {
        "qhf8xq": u1u1p7mj,
        "B5kzvoi": u8rtfz1,
        "oyh7mz": u1jmni06,
        "Bz10aip": u9yw1hi
      }
    }, indicatorProps === null || indicatorProps === void 0 ? void 0 : indicatorProps.sx]
  })))), indicatorPosition === 'outer-bottom' && /*#__PURE__*/React.createElement(CarouselIndicator, _extends({
    theme: "primary"
  }, indicatorProps)));
};
