// @universe-design/react/es/carousel/CarouselManager.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import produce from 'immer';
import clamp from 'lodash/clamp';
import isEqual from 'lodash/isEqual';
import minBy from 'lodash/minBy';
import { BehaviorSubject } from 'rxjs';
import { createContextHook } from '@universe-design/react-utilities';
import { forceReflow } from '../util/forceReflow';
import { getSlideFromAbsoluteIndex, getAbsoluteIndexFromSlide } from './transition-descriptors/utils';
export class CarouselManager {
  constructor(params) {
    // 类 prop 属性
    _defineProperty(this, "slideCount", 0);
    _defineProperty(this, "duration", 0);
    _defineProperty(this, "onSlideIndexChange", undefined);
    // subject
    _defineProperty(this, "_subject", void 0);
    _defineProperty(this, "_trackDescriptor", void 0);
    // 类 state 属性
    _defineProperty(this, "_prevSlide", null);
    _defineProperty(this, "_slide", void 0);
    _defineProperty(this, "_windowSize", void 0);
    _defineProperty(this, "_slideSizes", void 0);
    _defineProperty(this, "_slidesToBeRendered", void 0);
    // 最新值 cache
    _defineProperty(this, "_slideMetaCache", void 0);
    // eslint-disable-next-line func-call-spacing
    _defineProperty(this, "_transitionDisposers", new Map());
    _defineProperty(this, "_userEventLockSet", new Set());
    _defineProperty(this, "subscribe", onStoreChange => {
      const subscription = this._subject.subscribe({
        next: onStoreChange
      });
      return () => {
        subscription.unsubscribe();
      };
    });
    /**
     * 注册 slide 节点
     */
    _defineProperty(this, "register", slide => element => {
      if (this._getSlideMeta(slide)) {
        this._updateSlideMeta(slide, {
          element
        });
        return;
      }
      this._setSlideMeta(slide, {
        element
      });
    });
    this._trackDescriptor = params.trackDescriptor;
    this._slide = {
      groupIndex: 0,
      slideIndex: params.defaultSlideIndex
    };
    this._windowSize = {
      width: 0,
      height: 0
    };
    this._slideSizes = [];
    this._slideMetaCache = new Map();
    this._slidesToBeRendered = [];
    this._subject = new BehaviorSubject(this.state);
  }

  /**
   * ====================
   *   Public Methods
   * ====================
   */
  get state() {
    return {
      slide: this._slide,
      prevSlide: this._prevSlide,
      windowSize: this._windowSize,
      slideSizes: this._slideSizes,
      slidesToBeRendered: this._slidesToBeRendered
    };
  }
  get canUserInteract() {
    return this._userEventLockSet.size === 0;
  }
  /**
   * 更新视口尺寸
   */
  updateWindowSize(size) {
    // logger.info('manager.updateWindowSize', size);
    this._windowSize = produce(this._windowSize, draft => {
      const {
        width,
        height
      } = size;
      Object.assign(draft, {
        width,
        height
      });
    });
    this._emit();
  }

  /**
   * 更新 slide 尺寸
   */
  updateSlideSize(slide, size) {
    this._slideSizes = produce(this._slideSizes, draft => {
      const {
        width,
        height
      } = size;
      if (!draft[slide.slideIndex]) {
        draft[slide.slideIndex] = {
          width,
          height,
          mounted: true
        };
        return;
      }
      Object.assign(draft[slide.slideIndex], {
        width,
        height,
        mounted: true
      });
    });
    // logger.info('manager.updateSlideSize', slide, this._slideSizes);
    this._emit();
  }

  /**
   * 按 slide index 来设置当前 slide
   */
  setSlideIndex(slide) {
    let {
      triggerChange = true
    } = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    // logger.info('setSlideIndex', this._userEventLockSet);
    const _slide = clamp(slide, 0, this.slideCount - 1);
    const curAbsoluteIndex = getAbsoluteIndexFromSlide(this._slide, this.slideCount);
    const nextAbsoluteIndexes = [this._slide.groupIndex - 1, this._slide.groupIndex, this._slide.groupIndex + 1].map(groupIndex => getAbsoluteIndexFromSlide({
      groupIndex,
      slideIndex: _slide
    }, this.slideCount));
    const nextSlide = getSlideFromAbsoluteIndex(minBy(nextAbsoluteIndexes, i => Math.abs(i - curAbsoluteIndex)), this.slideCount);
    this._setSlide(nextSlide, {
      triggerChange
    });
  }

  /**
   * 根据步长来设置新的 slide
   */
  setSlideIndexByStep(step) {
    // logger.info('setSlideIndexByStep', this._userEventLockSet);
    const curAbsoluteIndex = getAbsoluteIndexFromSlide(this._slide, this.slideCount);
    const nextSlide = getSlideFromAbsoluteIndex(curAbsoluteIndex + step, this.slideCount);
    this._setSlide(nextSlide, {
      triggerChange: true
    });
  }

  /**
   * 更新渲染的 slides 列表
   */
  updateSlidesToBeRendered() {
    const nextSlides = this._trackDescriptor.getSlidesTOBeRendered(this);
    if (isEqual(this._slidesToBeRendered, nextSlides)) {
      return;
    }
    this._slidesToBeRendered = produce(this._slidesToBeRendered, draft => {
      draft.splice(nextSlides.length - 1, this._slidesToBeRendered.length - nextSlides.length);
      for (let i = 0; i < nextSlides.length; i++) {
        if (!draft[i]) {
          draft[i] = nextSlides[i];
          continue;
        }
        Object.assign(draft[i], nextSlides[i]);
      }
    });
    this._emit();
  }
  getSlideSize(slideIndex) {
    var _this$_slideSizes$sli;
    return (_this$_slideSizes$sli = this._slideSizes[slideIndex]) !== null && _this$_slideSizes$sli !== void 0 ? _this$_slideSizes$sli : {
      width: 0,
      height: 0
    };
  }

  // 1. 直接跳转到当前状态
  // 2. 渐变跳转到当前状态
  resetNodes(options) {
    for (const [key, {
      element
    }] of this._slideMetaCache) {
      const slide = this._unmarshalSlide(key);
      if (!element) {
        continue;
      }
      if (options.slick) {
        this._resetNodeSlickly(slide, element);
      } else {
        this._resetNode(slide, element);
      }
    }
  }

  /**
   * ====================
   *   Slide Meta Cache
   * ====================
   */

  // 获取 slide 的 cache key
  _marshalSlide(slide) {
    return JSON.stringify(slide);
  }
  _unmarshalSlide(key) {
    return JSON.parse(key);
  }
  // 获取 slide meta 信息
  _getSlideMeta(slide) {
    return this._slideMetaCache.get(this._marshalSlide(slide));
  }
  // 覆盖 slide meta 信息
  _setSlideMeta(slide, params) {
    return this._slideMetaCache.set(this._marshalSlide(slide), params);
  }
  // 更新 slide meta 信息
  _updateSlideMeta(slide, params) {
    const slideMeta = this._getSlideMeta(slide);
    if (!slideMeta) {
      return;
    }
    this._setSlideMeta(slide, {
      ...slideMeta,
      ...params
    });
  }

  /**
   * ====================
   *   State
   * ====================
   */
  _setSlide(nextSlide, options) {
    this._prevSlide = this._slide;
    this._slide = produce(this._slide, draft => {
      Object.assign(draft, nextSlide);
    });
    if (this._prevSlide !== this._slide && options.triggerChange) {
      var _this$onSlideIndexCha;
      (_this$onSlideIndexCha = this.onSlideIndexChange) === null || _this$onSlideIndexCha === void 0 ? void 0 : _this$onSlideIndexCha.call(this, this._slide.slideIndex);
    }
    this._emit();
  }
  _emit() {
    const nextState = produce(this._subject.value, draft => {
      Object.assign(draft, this.state);
    });
    if (this._subject.value !== nextState) {
      this._subject.next(this.state);
    }
  }
  _resetNode(slide, element) {
    this._trackDescriptor.onGotoDirectly({
      slide,
      element
    }, this);
  }
  _resetNodeSlickly(slide, element) {
    const cacheKey = this._marshalSlide(slide);
    // lock user event
    this._userEventLockSet.add(cacheKey);

    // 这里一定要有 disposer 是因为，如果旧的 transition 还未完成，新的 transition 就开始，这时旧的 transition 结束可能会影响新 transition 的工作
    const prevDisposer = this._transitionDisposers.get(cacheKey);
    if (prevDisposer) {
      prevDisposer();
    }

    // transition
    // enter
    this._trackDescriptor.onGotoSlicklyStart({
      slide,
      element
    }, this);
    // entering
    forceReflow(element);
    const enteringDisposer = this._trackDescriptor.onGotoSlicklyActive({
      slide,
      element
    }, this);
    // entered
    const t = window.setTimeout(() => {
      this._trackDescriptor.onGotoSlicklyDone({
        slide,
        element
      }, this);
      this._userEventLockSet.delete(cacheKey);
      // eslint-disable-next-line @typescript-eslint/no-use-before-define
      disposer(); // 这里 dispose 一下避免自定义渐变时可能存在的内存泄露
    }, this.duration);

    // set disposer
    const disposer = () => {
      enteringDisposer === null || enteringDisposer === void 0 ? void 0 : enteringDisposer();
      window.clearTimeout(t);
    };
    this._transitionDisposers.set(cacheKey, disposer);
  }
}
export const {
  Provider: CarouselManagerContextProvider,
  useContext: useCarouselManagerContext
} = createContextHook('CarouselManagerContext', {
  strict: true
});
