// @universe-design/react/es/carousel/CarouselSlide.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.u1l02sjl{height:100%;}
import udg2fzc from "./../.stylex/udg2fzc";
//.uly5x3f{width:100%;}
import u13d3rh8 from "./../.stylex/u13d3rh8";
//.u1euv43f{position:absolute;}
import u1u1p7mj from "./../.stylex/u1u1p7mj";
//.u1xynx9j{left:0px;}
import u184z4yl from "./../.stylex/u184z4yl";
//.u113wtx2{top:0px;}
import u1981n0b from "./../.stylex/u1981n0b";
//.u14t3ns0{display:inline-block;}
import u8wgnjj from "./../.stylex/u8wgnjj";
//.u1ewtqcl{box-sizing:border-box;}
import u1p85sbs from "./../.stylex/u1p85sbs";
//.uq62nq6{background-color:@text-disabled;}
import u19qsyxj from "./../.stylex/u19qsyxj";
import stylex from '@universe-design/stylex';
import * as React from 'react';
import { Transition } from 'react-transition-group';
import clsx from 'clsx';
import { useCarouselContext } from './CarouselContext';
import { useCarouselManagerContext } from './CarouselManager';
import { useResizeObserver } from '@universe-design/react-utilities';
import { getPositionAndSize, useIsomorphicLayoutEffect as useLayoutEffect
// useLogger
} from '../hooks';
import composeRefs from '../util/composeRefs';
import { forwardHTMLRef } from '../util/forwardRef';
import { createContextHook } from '@universe-design/react-utilities';
import pick from 'lodash/pick';
import omit from 'lodash/omit';
const transitionGroupClonePropKeys = ['onExited', 'in', 'appear', 'enter', 'exit'];
export const {
  Provider: CarouselSlideContextProvider,
  useContext: useCarouselSlideContext
} = createContextHook('CarouselSlideContext', {
  strict: true
});
/**
 * @public
 */
export const CarouselSlide = forwardHTMLRef((props, ref) => {
  const {
    children,
    className,
    fullHeight = true,
    fullWidth = true,
    ...restProps
  } = props;
  // https://github.com/reactjs/react-transition-group/blob/0d8a515aa59b8ba77fcabaac479e798532d4adfa/src/utils/ChildMapping.js#L88
  const transitionGroupProps = pick(restProps, transitionGroupClonePropKeys);
  const domAttrs = omit(restProps, transitionGroupClonePropKeys);
  const {
    groupIndex,
    slideIndex
  } = useCarouselSlideContext();
  const {
    duration,
    rootPrefixCls
  } = useCarouselContext();
  const manager = useCarouselManagerContext();
  const slideElRef = React.useRef(null);
  const slideElResizeObserverRef = useResizeObserver(entry => {
    const rect = getPositionAndSize(entry.target, entry);
    // logger.info(
    //   `Carousel useResizeObserver ${JSON.stringify({
    //     groupIndex,
    //     slideIndex,
    //   })}`,
    //   {
    //     width: rect.width,
    //     height: rect.height,
    //   }
    // );
    manager.updateSlideSize({
      groupIndex,
      slideIndex
    }, rect);
  });
  const composedSlideElRef = React.useMemo(() => composeRefs(ref, slideElRef, slideElResizeObserverRef, manager.register({
    groupIndex,
    slideIndex
  })), [groupIndex, manager, ref, slideElResizeObserverRef, slideIndex]);
  useLayoutEffect(() => {
    // 首次渲染立即存下 slide size
    const positionedEl = slideElRef.current;
    if (!positionedEl) {
      return;
    }
    // force reflow 但是没关系，因为都是 absolute 的元素
    manager.updateSlideSize({
      groupIndex,
      slideIndex
    }, getPositionAndSize(positionedEl));
  }, [groupIndex, manager, slideIndex]);

  // useLogger(`CarouselSlide ${JSON.stringify({ groupIndex, slideIndex })}`);

  return /*#__PURE__*/React.createElement(Transition, _extends({
    nodeRef: slideElRef,
    timeout: {
      exit: duration
    }
  }, transitionGroupProps), /*#__PURE__*/React.createElement("div", _extends({
    ref: composedSlideElRef,
    "data-group-index": groupIndex,
    "data-slide-index": slideIndex,
    className: clsx(`${rootPrefixCls}__slide`, stylex({
      "props": {
        "qhf8xq": u1u1p7mj,
        "oyh7mz": u184z4yl,
        "Bhzewxz": u1981n0b,
        "mc9l5x": u8wgnjj,
        "B7ck84d": u1p85sbs,
        "De3pzq": u19qsyxj
      }
    }, fullWidth && {
      "props": {
        "a9b677": u13d3rh8
      }
    }, fullHeight && {
      "props": {
        "Bqenvij": udg2fzc
      }
    }), className)
  }, domAttrs), children));
});

/**
 * @public
 * @docgen
 */
