// @universe-design/react/es/carousel/CarouselWindow.js
//.u1a3p1vp{overflow:hidden;}
import ux645fm from "./../.stylex/ux645fm";
//.uz5stix{white-space:nowrap;}
import uugkpzq from "./../.stylex/uugkpzq";
//.u10pi13n{position:relative;}
import u5jxv71 from "./../.stylex/u5jxv71";
//.uqerorx{flex-grow:1;}
import u99gbyh from "./../.stylex/u99gbyh";
import * as React from 'react';
import { TransitionGroup } from 'react-transition-group';
import { useSyncExternalStore } from 'use-sync-external-store/shim';
import clsx from 'clsx';
import stylex from '@universe-design/stylex';
import { getPositionAndSize, useIsomorphicLayoutEffect as useLayoutEffect } from '../hooks';
import { useResizeObserver } from '@universe-design/react-utilities';
import { useCarouselManagerContext } from './CarouselManager';
import { useCarouselContext } from './CarouselContext';
import { CarouselSlideContextProvider } from './CarouselSlide';
const CarouselSlideProvider = _ref => {
  let {
    children,
    slide,
    ...props
  } = _ref;
  return (
    /*#__PURE__*/
    // 这里用 ForwardProps 是因为 TransitionGroup 会给子元素 cloneProps，但是我们又需要给 list item 加 key
    React.createElement(CarouselSlideContextProvider, {
      value: slide
    }, /*#__PURE__*/React.cloneElement(children, props))
  );
};
export const CarouselWindow = _ref2 => {
  let {
    children
  } = _ref2;
  const {
    rootPrefixCls,
    duration,
    slide,
    slides,
    play,
    pause,
    autoPlayPauseOnHover
  } = useCarouselContext();
  const manager = useCarouselManagerContext();

  /**
   * ========================================
   *   Subscribe state changes
   * ========================================
   */
  const slidesToBeRendered = useSyncExternalStore(manager.subscribe, React.useCallback(() => manager.state.slidesToBeRendered, [manager]));
  const slideSizes = useSyncExternalStore(manager.subscribe, React.useCallback(() => manager.state.slideSizes, [manager]));
  const windowSize = useSyncExternalStore(manager.subscribe, React.useCallback(() => manager.state.windowSize, [manager]));

  /**
   * ========================================
   *   Window Size
   * ========================================
   */
  const ref = useResizeObserver(entry => {
    const rect = getPositionAndSize(entry.target, entry);
    manager.updateWindowSize(rect);
  });
  useLayoutEffect(() => {
    // 首次渲染
    const el = ref.current;
    if (!el) {
      return;
    }
    manager.updateWindowSize(getPositionAndSize(el));
  }, [manager, ref]);

  /**
   * ========================================
   *   Update Slides To Be Rendered
   * ========================================
   */
  useLayoutEffect(() => {
    manager.updateSlidesToBeRendered();
  });

  /**
   * ========================================
   *  Reset All Position Start
   * ========================================
   * TODO 这段逻辑很不优雅
   *
   * - slideIndex 变化以后需要重新做动画（中间几张图比较小，切换 slideIndex 后 slidesToBeRendered 没有变的情况）
   * - windowSize 和 slideSizes 的变化，当不引起 slidesToBeRendered 变化时也需要一步到位（视口大小变化，幻灯片尺寸变化，但是未引起 slidesToBeRendered 时）
   * - slidesToBeRendered 变化有两种情况
   *  - slideIndex 变化导致的 slidesToBeRendered 变化，prevSlides 和 nextSlides 需要一起做动画
   *  - windowSize 和 slideSizes 导致的 slidesToBeRendered 变化，需要立即到位
   *
   */
  const changed = React.useRef({
    slide: false,
    windowSize: false,
    slideSizes: false
  });
  useLayoutEffect(() => {
    // 这个逻辑是保证 slidesToBeRendered windowSize 变化也能触发直接跳转

    changed.current.windowSize = true;
    manager.resetNodes({
      slick: false
    });
  }, [manager, windowSize]);
  useLayoutEffect(() => {
    // 这个逻辑是保证 slidesToBeRendered 不变时，slideSizes 变化也能触发直接跳转

    changed.current.slideSizes = true;
    manager.resetNodes({
      slick: false
    });
  }, [manager, slideSizes]);
  useLayoutEffect(() => {
    // 这个逻辑是保证 slidesToBeRendered 不变时，slide 变化也能触发动画跳转
    const el = ref.current;
    if (!el) {
      return;
    }
    slide.groupIndex;
    slide.slideIndex;
    changed.current.slide = true;
    // 这里所有的 size 都已经稳定，所有 slides 都已经存在，重置一次位置，并开始做动画
    manager.resetNodes({
      slick: true
    });
  }, [duration, manager, ref, slide.groupIndex, slide.slideIndex]);
  useLayoutEffect(() => {
    // 由于 updateSlidesToBeRendered 是在上面几个 change 过程的下一个 render 周期才会做出反应，所以需要判断下 slidesToBeRendered 变化是什么导致的，并做出对应的措施。
    const el = ref.current;
    if (!el) {
      return;
    }
    if (changed.current.slide) {
      changed.current.slide = false;
      changed.current.windowSize = false;
      changed.current.slideSizes = false;
      manager.resetNodes({
        slick: true
      });
    }
    if (changed.current.windowSize || changed.current.slideSizes) {
      changed.current.windowSize = false;
      changed.current.slideSizes = false;
      manager.resetNodes({
        slick: false
      });
    }
  }, [duration, manager, ref, slidesToBeRendered]);
  /**
   * ========================================
   *  Reset All Position End
   * ========================================
   */

  // useLogger('CarouselWindow', { slidesToBeRendered, slideSizes, slide });

  return /*#__PURE__*/React.createElement("div", {
    ref: ref,
    className: clsx(`${rootPrefixCls}__window`, stylex({
      "props": {
        "Bpg54ce": ux645fm,
        "Huce71": uugkpzq,
        "qhf8xq": u5jxv71,
        "Bh6795r": u99gbyh
      }
    })),
    onMouseEnter: () => {
      if (!autoPlayPauseOnHover) {
        return;
      }
      pause();
    },
    onMouseLeave: () => {
      play();
    }
  }, /*#__PURE__*/React.createElement(TransitionGroup, {
    className: `${rootPrefixCls}__track`
  }, slidesToBeRendered.map(_slide => /*#__PURE__*/React.createElement(CarouselSlideProvider, {
    key: `${_slide.groupIndex}-${_slide.slideIndex}`,
    slide: _slide
  }, slides[_slide.slideIndex]))), children);
};
