// @universe-design/react/es/carousel/CarouselIndicator.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.u1euv43f{position:absolute;}
import u1u1p7mj from "./../.stylex/u1u1p7mj";
//.uclvdme{left:-3px;}
import u14zp1q4 from "./../.stylex/u14zp1q4";
//.uvq7pei{right:-3px;}
import u1ska7bq from "./../.stylex/u1ska7bq";
//.u12v7yn{top:-3px;}
import umr338j from "./../.stylex/umr338j";
//.uode0jj{bottom:-3px;}
import u17syspr from "./../.stylex/u17syspr";
//.u16dn6v3{width:6px;}
import u165j2wq from "./../.stylex/u165j2wq";
//.uqecqfw{background-color:@static-white-30;}
import uvfwkig from "./../.stylex/uvfwkig";
//.u12w28jm:hover{background-color:@static-white-60;}
import ud7fs9g from "./../.stylex/ud7fs9g";
//.u1qhvef2{background-color:@line-border-component;}
import u1d4y5un from "./../.stylex/u1d4y5un";
//.u1qh4al8:hover{background-color:@primary-pri-200;}
import uka18dz from "./../.stylex/uka18dz";
//.uq4mcun{width:24px;}
import uymufub from "./../.stylex/uymufub";
//.u6ywr7j{height:4px;}
import u9dhz2c from "./../.stylex/u9dhz2c";
//.urx94fk{width:12px;}
import u6rgktp from "./../.stylex/u6rgktp";
//.u3mu39s{height:6px;}
import ub4ucl8 from "./../.stylex/ub4ucl8";
//.u1su79po{background-color:@primary-on-primary-fill;}
import uq3qnqa from "./../.stylex/uq3qnqa";
//.u1oquvwe{background-color:@primary-fill-default;}
import u15s3zb6 from "./../.stylex/u15s3zb6";
//.u10pi13n{position:relative;}
import u5jxv71 from "./../.stylex/u5jxv71";
//.u7o03ib{transition-property:all;}
import u18eu6n2 from "./../.stylex/u18eu6n2";
//.u1k6fduh{cursor:pointer;}
import u12o6wmt from "./../.stylex/u12o6wmt";
//.u4d9j23{justify-content:center;}
import ut7q8es from "./../.stylex/ut7q8es";
//.uc7z3ec{padding-top:20px;}
import u1xft5cu from "./../.stylex/u1xft5cu";
//.uhxju0i{padding-right:0;}
import u11atsnw from "./../.stylex/u11atsnw";
//.ue2my4m{padding-bottom:20px;}
import u1co87pz from "./../.stylex/u1co87pz";
//.u1cnd47f{padding-left:0;}
import uhzi0ua from "./../.stylex/uhzi0ua";
//.ui64zpg{flex-shrink:0;}
import uidkx17 from "./../.stylex/uidkx17";
import * as React from 'react';
import stylex from '@universe-design/stylex';
import { forwardHTMLRef } from '../util/forwardRef';
import { useCarouselContext } from './CarouselContext';
import { sxBorderRadius, sxTransitionTimingFunction } from '../stylex';
import { Space } from '../space';
import clsx from 'clsx';
import { useCarouselManagerContext } from './CarouselManager';

/**
 * @public
 * @docgen
 */

export const CarouselIndicator = forwardHTMLRef((props, ref) => {
  const {
    className,
    type = 'dot',
    theme = 'light',
    sx,
    ...restProps
  } = props;
  const {
    slide,
    slideCount,
    duration,
    rootPrefixCls
  } = useCarouselContext();
  const manager = useCarouselManagerContext();
  return /*#__PURE__*/React.createElement(Space, _extends({
    className: clsx(`${rootPrefixCls}__indicator`, stylex({
      "props": {
        "Brf1p80": ut7q8es,
        "z8tnut": u1xft5cu,
        "z189sj": u11atsnw,
        "Byoj8tv": u1co87pz,
        "uwmqm3": uhzi0ua,
        "Bnnss6s": uidkx17
      }
    }, sx), className),
    size: 6,
    ref: ref,
    inline: false
  }, restProps), Array(slideCount).fill(0).map((_, i) => /*#__PURE__*/React.createElement("div", {
    key: i,
    style: {
      transitionDuration: `${duration}ms`
    },
    className: clsx(`${rootPrefixCls}__indicator-item`, `${rootPrefixCls}__indicator-item-${theme}`, `${rootPrefixCls}__indicator-item-${type}`, slide.slideIndex === i && `${rootPrefixCls}__indicator-item-active`, stylex({
      "props": {
        "qhf8xq": u5jxv71,
        "Bmy1vo4": u18eu6n2,
        "Bceei9c": u12o6wmt
      }
    }, sxBorderRadius.xl, sxTransitionTimingFunction.standard, slide.slideIndex === i ? [theme === 'primary' ? {
      "props": {
        "De3pzq": u15s3zb6
      }
    } : {
      "props": {
        "De3pzq": uq3qnqa
      }
    }, type === 'dot' ? {
      "props": {
        "a9b677": u6rgktp,
        "Bqenvij": ub4ucl8
      }
    } : {
      "props": {
        "a9b677": uymufub,
        "Bqenvij": u9dhz2c
      }
    }] : [theme === 'primary' ? {
      "props": {
        "De3pzq": u1d4y5un,
        "Jwef8y": uka18dz
      }
    } : {
      "props": {
        "De3pzq": uvfwkig,
        "Jwef8y": ud7fs9g
      }
    }, type === 'dot' ? {
      "props": {
        "a9b677": u165j2wq,
        "Bqenvij": ub4ucl8
      }
    } : {
      "props": {
        "a9b677": uymufub,
        "Bqenvij": u9dhz2c
      }
    }])),
    onClick: () => {
      if (!manager.canUserInteract) {
        return;
      }
      manager.setSlideIndex(i);
    }
  }, /*#__PURE__*/React.createElement("div", {
    className: stylex({
      "props": {
        "qhf8xq": u1u1p7mj,
        "oyh7mz": u14zp1q4,
        "j35jbq": u1ska7bq,
        "Bhzewxz": umr338j,
        "B5kzvoi": u17syspr
      }
    })
  }))));
});
