// @universe-design/react/es/drawer/Drawer.js
import _extends from "@babel/runtime/helpers/esm/extends";
/* eslint-disable @typescript-eslint/no-magic-numbers */
import React, { useCallback } from 'react';
import stylex from '@universe-design/stylex';
import CloseOutlined from '@universe-design/icons-react/CloseOutlined';
import isNil from 'lodash/isNil';
import clsx from 'clsx';
import { Transition } from 'react-transition-group';
import Portal from '../portal';
import { Button } from '../button';
import { useCssScope } from '../config-provider/hooks';
import { FocusLock } from '../shared';
import { getDefaultSizeAbbr, isFunction } from '../util';
import { useAnimatedVisible } from '../shared/components/transition/useAnimatedVisible';
import { FadeTransition } from '../shared/components/transition/FadeTransition';
import { SlideTransition } from '../shared/components/transition/SlideTransition';
import { cssTypographyBody0, cssTypographyTitle4 } from '../stylex';
import './style';

/**
 * @public
 */

// TODO: 1. warn 换 utils 内的 2.考虑放到 utils 下
function compatibleDeprecatedProps(props, key, deprecatedKey) {
  if (!isNil(props[key])) {
    return props[key];
  }
  if (!isNil(props[deprecatedKey])) {
    console.warn(`"${String(deprecatedKey)}" is deprecated, please use "${String(key)}" instead.`);
    return props[deprecatedKey];
  }
}
const toSlideDirectionMap = {
  left: 'right',
  right: 'left',
  bottom: 'up',
  top: 'down'
};
const useDrawerCssScope = () => useCssScope('drawer');
const getNumericalSize = size => {
  switch (size) {
    case 'large':
      return 680;
    case 'medium':
      return 480;
    case 'small':
      return 350;
    default:
      throw new Error(`Size \`${size}\` is not support.`);
  }
};
export const Drawer = props => {
  const {
    visible = false,
    placement = 'right',
    size = 'small',
    width = getNumericalSize(size),
    height = getNumericalSize(size),
    style: wrapperStyle,
    maskStyle,
    mask = true,
    maskClosable = true,
    keyboard = true,
    destroyOnClose = false,
    className: wrapperClassName,
    bodyPaddingSize = 'medium',
    prefixCls,
    zIndex,
    forceRender,
    getContainer,
    afterVisibleChange,
    onClose,
    footer,
    children,
    initialFocusRef,
    finalFocusRef,
    enableFocusLock = false,
    enableRestoreFocus = false
  } = props;
  const showCloseIcon = compatibleDeprecatedProps(props, 'showCloseIcon', 'closable');
  const defaultScope = useDrawerCssScope();
  const cssScope = prefixCls || defaultScope();
  const timeout = React.useMemo(() => ({
    appear: 300,
    enter: 300,
    exit: 200
  }), []);
  const {
    transitionGroupProps,
    transitionGroupPropsWithoutListener,
    styleDisplay,
    animatedVisible,
    noopRef
  } = useAnimatedVisible({
    destroyOnClose,
    forceRender,
    visible,
    onAfterClose: React.useCallback(() => isFunction(afterVisibleChange) && afterVisibleChange(false), [afterVisibleChange]),
    onAfterOpen: React.useCallback(() => isFunction(afterVisibleChange) && afterVisibleChange(true), [afterVisibleChange])
  });
  const slideExtraProps = React.useMemo(() => {
    var _ref;
    const direction = toSlideDirectionMap[placement];
    return {
      direction,
      size: (_ref = ['left', 'right'].includes(direction) ? width : height) !== null && _ref !== void 0 ? _ref : 0
    };
  }, [height, placement, width]);
  const renderCloseIcon = () => {
    const {
      closeIcon = /*#__PURE__*/React.createElement(Button, {
        icon: /*#__PURE__*/React.createElement(CloseOutlined, null),
        size: "medium",
        type: "icon"
      })
    } = props;
    return /*#__PURE__*/React.createElement("div", {
      className: `${cssScope}__close`,
      onClick: onClose
    }, closeIcon);
  };
  const renderHeader = () => {
    const {
      title,
      subTitle,
      headerExtra,
      headerStyle
    } = props;
    // compatible other DS's props
    if (!title && !showCloseIcon) {
      return null;
    }
    return /*#__PURE__*/React.createElement("div", {
      className: clsx(`${cssScope}__header`),
      style: headerStyle
    }, /*#__PURE__*/React.createElement("div", {
      className: `${cssScope}__header__content`
    }, (title || subTitle) && /*#__PURE__*/React.createElement("div", {
      className: clsx(`${cssScope}__header__title`, stylex(cssTypographyTitle4))
    }, title), showCloseIcon && renderCloseIcon()), /*#__PURE__*/React.createElement("div", {
      className: `${cssScope}__header__extra`
    }, subTitle && /*#__PURE__*/React.createElement("p", {
      className: clsx(`${cssScope}__header__subTitle`, stylex(cssTypographyBody0))
    }, subTitle), headerExtra ? headerExtra : null));
  };
  const renderFooter = () => {
    const {
      footerStyle
    } = props;
    if (!footer) {
      return null;
    }
    return /*#__PURE__*/React.createElement("div", {
      className: `${cssScope}__footer`,
      style: footerStyle
    }, footer);
  };
  const onMaskClick = useCallback(e => {
    if (!maskClosable) {
      return;
    }
    if (onClose) {
      onClose(e);
    }
  }, [maskClosable, onClose]);
  const maskRef = React.useRef(null);
  const renderMask = () => {
    if (!mask) {
      return;
    }
    const mergeMaskStyle = {
      ...maskStyle,
      ...styleDisplay
    };
    if (zIndex || zIndex === 0) {
      mergeMaskStyle.zIndex = zIndex;
    }
    return /*#__PURE__*/React.createElement(FadeTransition, _extends({
      timeout: timeout,
      appear: true,
      nodeRef: maskRef
    }, transitionGroupPropsWithoutListener), /*#__PURE__*/React.createElement("div", {
      ref: maskRef,
      className: `${cssScope}__mask`,
      style: mergeMaskStyle,
      onClick: onMaskClick,
      tabIndex: -1
    }));
  };
  const onKeyDown = useCallback(e => {
    if (!keyboard) {
      return;
    }
    // ESC

    if (e.keyCode === 27) {
      e.stopPropagation();
      if (onClose) {
        onClose(e);
      }
    }
  }, [onClose, keyboard]);
  const dialogRef = React.useRef(null);
  const mergedInitialFocusRef = initialFocusRef === null ? null : initialFocusRef !== null && initialFocusRef !== void 0 ? initialFocusRef : dialogRef;
  const renderContent = () => {
    const {
      bodyStyle: bodyStyleProps,
      drawerStyle
    } = props;
    const contentStyle = {
      ...drawerStyle,
      ...styleDisplay,
      ...(height && ['top', 'bottom'].includes(placement) ? {
        height
      } : {}),
      ...(width && ['left', 'right'].includes(placement) ? {
        width
      } : {}),
      zIndex
    };
    const bodyStyle = {
      ...bodyStyleProps,
      ...(typeof bodyPaddingSize === 'number' ? {
        padding: `${bodyPaddingSize}px`
      } : {})
    };
    const bodyPaddingKeywords = typeof bodyPaddingSize !== 'number' ? getDefaultSizeAbbr(bodyPaddingSize) : '';
    const mergedEnableFocusLock = (() => {
      // 在非 mask 情况, 即使用户没有传 enableFocusLock, 也要关掉lock, 否则按照用户传的为准
      if (!mask && !enableFocusLock) {
        return false;
      }
      return enableFocusLock;
    })();
    return /*#__PURE__*/React.createElement(SlideTransition, _extends({
      appear: true,
      timeout: timeout,
      extraProps: slideExtraProps,
      nodeRef: dialogRef
    }, transitionGroupPropsWithoutListener), /*#__PURE__*/React.createElement(FocusLock, {
      isDisabled: !visible,
      initialFocusRef: mergedInitialFocusRef,
      restoreFocus: enableRestoreFocus,
      finalFocusRef: finalFocusRef,
      enableFocusLock: mergedEnableFocusLock
    }, /*#__PURE__*/React.createElement("div", {
      ref: dialogRef,
      role: "dialog",
      "aria-modal": "true",
      tabIndex: -1,
      onKeyDown: onKeyDown,
      className: clsx(`${cssScope}__content__wrapper`),
      style: contentStyle
    }, renderHeader(), /*#__PURE__*/React.createElement("div", {
      className: clsx(`${cssScope}__content`, {
        [`${cssScope}__content__nofooter`]: footer,
        [`${cssScope}__content__pd-${bodyPaddingKeywords}`]: bodyPaddingKeywords
      }),
      style: bodyStyle
    }, children), renderFooter())));
  };
  return /*#__PURE__*/React.createElement(Transition, _extends({
    nodeRef: noopRef,
    timeout: timeout,
    appear: true
  }, transitionGroupProps), (() => {
    const childNodes = /*#__PURE__*/React.createElement("div", {
      className: clsx(`${cssScope}__root`, {
        [`${cssScope}-${placement}`]: true,
        [`${cssScope}-open`]: animatedVisible
      }),
      style: wrapperStyle
    }, renderMask(), renderContent());
    if (getContainer === false) {
      return /*#__PURE__*/React.createElement("div", {
        className: wrapperClassName
      }, childNodes);
    }
    return /*#__PURE__*/React.createElement(Portal, {
      wrapperClassName: wrapperClassName,
      container: getContainer
    }, childNodes);
  })());
};
