// @universe-design/react/lib/pin-input/usePinInput.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.usePinInputField = exports.usePinInputDescendantsContext = exports.usePinInputDescendants = exports.usePinInputDescendant = exports.usePinInputContext = exports.usePinInput = exports.PinInputDescendantsProvider = exports.PinInputContextProvider = void 0;
var _u1wa1nxx = _interopRequireDefault(require("./../.stylex/u1wa1nxx"));
var _u1p85sbs = _interopRequireDefault(require("./../.stylex/u1p85sbs"));
var _u19bueuo = _interopRequireDefault(require("./../.stylex/u19bueuo"));
var _u1ust6hi = _interopRequireDefault(require("./../.stylex/u1ust6hi"));
var _u1cpcpb = _interopRequireDefault(require("./../.stylex/u1cpcpb5"));
var _u10qn3go = _interopRequireDefault(require("./../.stylex/u10qn3go"));
var _uflkg = _interopRequireDefault(require("./../.stylex/uflkg16"));
var _uvn5txf = _interopRequireDefault(require("./../.stylex/uvn5txf"));
var _us9j13g = _interopRequireDefault(require("./../.stylex/us9j13g"));
var _umt7kug = _interopRequireDefault(require("./../.stylex/umt7kug"));
var _uh3lwu = _interopRequireDefault(require("./../.stylex/uh3lwu4"));
var _ur = _interopRequireDefault(require("./../.stylex/ur61548"));
var _u1ul1s2h = _interopRequireDefault(require("./../.stylex/u1ul1s2h"));
var _u7vf9al = _interopRequireDefault(require("./../.stylex/u7vf9al"));
var _up3vbyy = _interopRequireDefault(require("./../.stylex/up3vbyy"));
var _react = _interopRequireWildcard(require("react"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _clsx = _interopRequireDefault(require("clsx"));
var _hooks = require("../hooks");
var _reactUtilities = require("@universe-design/react-utilities");
var _size = require("../util/size");
var _util = require("../util");
var _function = require("../util/function");
var _composeRefs = _interopRequireDefault(require("../util/composeRefs"));
var _configProvider = require("../config-provider");
var _stylex2 = require("../stylex");
var _sx = require("./sx");
require("./style");
//.ur7e4de{caret-color:transparent;}
//.u1ewtqcl{box-sizing:border-box;}
//.uses1vf{font-size:14px;}
//.u17mccla{text-align:center;}
//.u5ogflp{border-top-width:1px;}
//.u1hqa2wf{border-right-width:1px;}
//.u1f09k3d{border-bottom-width:1px;}
//.uinvdd3{border-left-width:1px;}
//.uzkkow9{border-top-style:solid;}
//.ucdblym{border-right-style:solid;}
//.ug706s2{border-bottom-style:solid;}
//.ujik90z{border-left-style:solid;}
//.u2hkw1w:focus-visible{outline-style:none;}
//.u2ys1bh{transition-property:border-color;}
//.u15vu5tu{transition-duration:0.1s;}
/**
 * ================================
 *
 * Create context to track descendants and their indices
 *
 * ================================
 */
const [PinInputDescendantsProvider, usePinInputDescendantsContext, usePinInputDescendants, usePinInputDescendant] = (0, _reactUtilities.createDescendantContext)();

/**
 * ================================
 *
 * Create Context
 *
 * ================================
 */
exports.usePinInputDescendant = usePinInputDescendant;
exports.usePinInputDescendants = usePinInputDescendants;
exports.usePinInputDescendantsContext = usePinInputDescendantsContext;
exports.PinInputDescendantsProvider = PinInputDescendantsProvider;
const {
  Provider: PinInputContextProvider,
  useContext: usePinInputContext
} = (0, _reactUtilities.createContextHook)('PinInputContext', {
  strict: true
});

/**
 * ================================
 *
 * Pin Input
 *
 * ================================
 */

/**
 * @public
 */
exports.usePinInputContext = usePinInputContext;
exports.PinInputContextProvider = PinInputContextProvider;
const toCharArray = value => value.split('');
const validate = (v, type) => {
  let regex = /^[0-9]+$/;
  if (type === 'alphanumeric') {
    regex = /^[0-9a-zA-Z]+$/;
  }
  if (type === 'alphabetic') {
    regex = /^[a-zA-Z]+$/;
  }
  return regex.test(v);
};
/**
 * 用于判断当前应该在哪个 input 中输入，规则是第一个空值的 input
 */
const getShouldInputIndex = (values, descendants) => {
  for (let i = 0; i < values.length; i++) {
    if (!values[i]) {
      return i;
    }
  }
  return Math.min(values.length, descendants.count() - 1);
};
const usePinInput = (props, ref) => {
  const {
    autoFocus = false,
    value,
    defaultValue = '',
    size = 'medium',
    onChange,
    onComplete,
    placeholder = '',
    id,
    disabled = false,
    error = false,
    mask = false,
    type = 'number'
  } = props;
  const rootCssScope = (0, _configProvider.useCssScope)('pin-input');
  const descendants = usePinInputDescendants();
  const [focusedIndex, setFocusedIndex] = _react.default.useState(-1);
  const [values, setValues] = (0, _reactUtilities.useControlledState)(
  // defaultValue 需要去除所有的非法字符
  toCharArray(defaultValue).filter(v => validate(v, type)) || [],
  // value
  (0, _util.isNonNullable)(value) ? toCharArray(value).filter(v => validate(v, type)) : [],
  // options
  {
    onChange: _react.default.useCallback(nextValues => {
      onChange === null || onChange === void 0 ? void 0 : onChange(nextValues.join(''));
      // complete 放在 change 后面
      const isComplete = nextValues.every(Boolean) && nextValues.length === descendants.count();
      if (isComplete) {
        onComplete === null || onComplete === void 0 ? void 0 : onComplete(nextValues.join(''));
      }
    }, [descendants, onChange, onComplete]),
    isControlledOverride: 'value' in props
  });

  /**
   * 设置下一个 input 为聚焦状态
   */
  const focusNext = _react.default.useCallback(index => {
    const next = descendants.next(index, false);
    if (next) {
      (0, _util.focus)(next.node, {
        selectTextIfInput: false
      });
    }
  }, [descendants]);
  const focusByIndex = _react.default.useCallback((index, options) => {
    const descendant = descendants.values()[index];
    if (descendant) {
      (0, _util.focus)(descendant.node, {
        selectTextIfInput: false,
        ...options
      });
    }
  }, [descendants]);

  /**
   * 设置 input 的单字符 value
   */
  const setValue = _react.default.useCallback((newValue, index) => {
    const nextValues = [...values];
    nextValues[index] = newValue;
    focusNext(index);
    setValues(nextValues);
  }, [values, focusNext, setValues]);

  /**
   * 重置 pin-input 输入框内容，并 focus 第一个 input
   */
  const clear = _react.default.useCallback(() => {
    const nextValues = Array(descendants.count()).fill('');
    setValues(nextValues);
    const first = descendants.first();
    if (first) {
      (0, _util.focus)(first.node, {
        selectTextIfInput: false
      });
    }
  }, [descendants, setValues]);

  /**
   * autoFocus 逻辑，只生效一次，后续的变化都不触发
   */
  (0, _hooks.useEffectOnce)(() => {
    if (autoFocus) {
      // useEffect 是比 useLayoutEffect 里使用 requestAnimation 和 setTimeout 调用时机都要早
      focusByIndex(getShouldInputIndex(values, descendants));
    }
  });
  (0, _react.useImperativeHandle)(ref, () => ({
    clear,
    focus: option => {
      focusByIndex(getShouldInputIndex(values, descendants), option);
    }
  }), [clear, descendants, focusByIndex, values]);
  return {
    descendants,
    rootCssScope,
    // props
    id,
    size,
    placeholder,
    disabled,
    error,
    mask,
    // states
    values,
    focusedIndex,
    // actions
    setValue,
    setValues,
    clear,
    setFocusedIndex,
    focusByIndex,
    type
  };
};

/**
 * ================================
 *
 * Pin Input Field
 *
 * ================================
 */

/**
 * @public
 */
exports.usePinInput = usePinInput;
const sxPinInputField = {
  root: _stylex.default.compose({
    "props": {
      "B7ck84d": _u1p85sbs.default,
      "Be2twd7": _u19bueuo.default,
      "fsow6f": _u1ust6hi.default,
      "B4j52fo": _u1cpcpb.default,
      "Bekrc4i": _u10qn3go.default,
      "Bn0qgzm": _uflkg.default,
      "ibv6hh": _uvn5txf.default,
      "icvyot": _us9j13g.default,
      "vrafjx": _umt7kug.default,
      "oivjwe": _uh3lwu.default,
      "wvpqe5": _ur.default,
      "B486eqv": _u1ul1s2h.default,
      "Bmy1vo4": _u7vf9al.default,
      "B3o57yi": _up3vbyy.default
    }
  }, _stylex2.sxBorderRadius.sm, _stylex2.sxTransitionTimingFunction.standard),
  small: _stylex.default.compose(_stylex2.sxInput.small, _stylex2.sxInputWidth.small),
  medium: _stylex.default.compose(_stylex2.sxInput.medium, _stylex2.sxInputWidth.medium),
  large: _stylex.default.compose(_stylex2.sxInput.large, _stylex2.sxInputWidth.large),
  complete: {
    "props": {
      "pb3013": _u1wa1nxx.default
    }
  }
};
const usePinInputField = (_ref, ref) => {
  let {
    sx,
    ...props
  } = _ref;
  const {
    setValue,
    setValues,
    focusByIndex,
    setFocusedIndex,
    focusedIndex,
    mask,
    disabled,
    id,
    values,
    placeholder,
    size,
    rootCssScope,
    error,
    type
  } = usePinInputContext();
  const {
    index,
    descendants,
    register
  } = usePinInputDescendant();
  const inputRef = _react.default.useRef(null);
  const latestValues = (0, _reactUtilities.useConstant)(() => ({
    get index() {
      var _inputRef$current;
      // https://github.com/chakra-ui/chakra-ui/blob/main/packages/descendant/src/use-descendant.ts#L59
      return (_inputRef$current = inputRef.current) !== null && _inputRef$current !== void 0 && _inputRef$current.dataset.index ? Number(inputRef.current.dataset.index) : -1;
    }
  }));
  const isComplete = Boolean(index === descendants.count() - 1 && values[index]);
  const inputProps = {
    'aria-label': 'Please enter your pin code',
    inputMode: 'numeric',
    type: mask ? 'password' : 'tel',
    id: (0, _util.isNonNullable)(id) ? `${id}-${index}` : undefined,
    disabled,
    value: values[index] || '',
    placeholder: focusedIndex === index ? '' : placeholder,
    ref: (0, _composeRefs.default)(ref, inputRef, register),
    autoComplete: 'off',
    tabIndex: getShouldInputIndex(values, descendants) === index ? undefined : -1,
    // TODO 由于 descendant register 不会触发其他参与项目的 rerender。这个值在 input field 动态变化的时候不一定准，例如，一个 4 个 field 当前 focus 到 index 3，此时，在不改变 focusIndex 的情况下修改 field 为 3 个，这时就完全无法 tab 到 pin-input 上
    ...props,
    className: (0, _clsx.default)(rootCssScope('-field'), rootCssScope(`-field-${(0, _size.getSizeAbbr)(size)}`), focusedIndex >= 0 && focusedIndex >= index && rootCssScope('-field-active'), error && rootCssScope('-field-error'), disabled && rootCssScope('-field-disabled'), (0, _stylex.default)(...(sx !== null && sx !== void 0 ? sx : [sxPinInputField.root, sxPinInputField[size], isComplete && sxPinInputField.complete, index !== 0 && _sx.sxPinInputMargin[size]])), props.className),
    style: {
      ...(mask ? {
        fontFamily: 'Tahoma, PingFang SC, Microsoft Yahei' // https://codesandbox.io/s/infallible-feather-j4oez3?file=/src/App.tsx 在测试中发现 tahoma 是 win 和 mac 都系统自带的效果最好的字体，剩下两个是各自系统默认
      } : {}),
      ...props.style
    },
    onChange: (0, _function.composeEventHandlers)(props.onChange, event => {
      const eventValue = event.currentTarget.value;
      // input 中的 value 被删除
      if (eventValue === '') {
        setValue('', index);
        return;
      }

      // 输入值不合法
      if (!validate(eventValue, type)) {
        return;
      }

      // 普通输入元素的时候
      if (eventValue.length === 1) {
        setValue(eventValue, index);
        return;
      }

      // 如果是最后一个元素，输入 2 个字符时需要覆盖当前值
      if (index === descendants.count() - 1) {
        setValue(eventValue[1], index);
        return;
      }

      // 复制粘贴的时候
      const nextValues = values.slice(0, index).concat(eventValue.split('')).filter((_, i) => i < descendants.count());
      focusByIndex(nextValues.length - 1);
      setValues(nextValues);
    }),
    onKeyDown: (0, _function.composeEventHandlers)(props.onKeyDown, event => {
      if (event.key === _reactUtilities.KeyboardKey.Backspace) {
        if (event.currentTarget.value === '') {
          const preInput = descendants.prev(index, false);
          if (preInput) {
            setValue('', index - 1);
            (0, _util.focus)(preInput.node, {
              selectTextIfInput: false
            });
          }
        } else {
          setValue('', index);
        }
      }
    }),
    onMouseDown: (0, _function.composeEventHandlers)(props.onMouseDown, ev => {
      // 如果当前 input 不该被 focus 就 preventDefault
      const shouldInputIndex = getShouldInputIndex(values, descendants);
      if (shouldInputIndex !== index) {
        ev.preventDefault();
        focusByIndex(shouldInputIndex);
      }
    }),
    onFocusChange: (focused, ev) => {
      var _props$onFocusChange;
      (_props$onFocusChange = props.onFocusChange) === null || _props$onFocusChange === void 0 ? void 0 : _props$onFocusChange.call(props, focused, ev);
      if (focused) {
        // autoFocus 的时候 index 还没有更新，但是 dom 上的 index 已经更新了，所以需要从 dom 上取 index。
        setFocusedIndex(latestValues.index);
      }
    },
    onBlur: (0, _function.composeEventHandlers)(props.onBlur, () => {
      setFocusedIndex(-1);
    })
  };
  return {
    inputProps
  };
};
exports.usePinInputField = usePinInputField;
