// @universe-design/react/src/dialog/stories/homeless.stories.tsx
/** @jsxImportSource @emotion/react */
import BanOutlined from '@universe-design/icons-react/BanOutlined';
import type { FC } from 'react';
import React, { useState } from 'react';
import type { ModalProps } from '..';
import { useDialog } from '..';
import { Button, Dialog, Divider, Drawer, Form, Menu, Popover, Radio, Select, Space, contextMenu } from '../../';
import { useFirstMountState } from '../../hooks';

export default {
  title: 'Components/Dialog/临时示例',
  component: Dialog,
};

type Base = ModalProps & { text?: string };

const RenderBase: FC<Base> = props => {
  const { text = '点我', visible: visibleInProps, ...modalProps } = props;
  const [visible, setVisible] = useState<boolean>(Boolean(visibleInProps));
  return (
    <div style={{ display: 'inline-block' }}>
      <Dialog
        title="标题"
        description="这是描述这是描述这是描述这是描述这是描述这是描述这是描述这是描述这是描述这是描述"
        visible={visible}
        centered
        destroyOnClose
        onOk={() => setVisible(false)}
        onCancel={() => setVisible(false)}
        {...modalProps}
      >
        <div>
          这是内容部分，这是内容部分这是内容部分这是内容部分这是内容部分这是内容部分这是内容部分这是内容部分这是内容部分这是内容部分这是内容部分
        </div>
      </Dialog>
      <Button style={{ marginRight: 10 }} onClick={() => setVisible(true)} type="primary">
        {text}
      </Button>
    </div>
  );
};

const RenderLoading: FC<any> = () => {
  const [visible, setVisible] = React.useState<boolean>(false);
  const [loading, setLoading] = React.useState<boolean>(false);
  return (
    <div style={{ display: 'inline-block' }}>
      <Dialog
        title="标题 - 点击确认后等待 3s 后关闭"
        description="这是描述这是描述这是描述这是描述这是描述这是描述这是描述这是描述这是描述这是描述"
        visible={visible}
        centered
        destroyOnClose
        confirmLoading={loading}
        onOk={() => {
          setLoading(true);
          setTimeout(() => {
            setLoading(false);
            setVisible(false);
          }, 3000);
        }}
        onCancel={() => setVisible(false)}
      >
        <div>
          这是内容部分，这是内容部分这是内容部分这是内容部分这是内容部分这是内容部分这是内容部分这是内容部分这是内容部分这是内容部分这是内容部分
        </div>
      </Dialog>
      <Button onClick={() => setVisible(true)}>点击确认后 等待3s </Button>
    </div>
  );
};

const RenderNoFooter: FC<any> = () => {
  const [visible, setVisible] = React.useState<boolean>(false);
  return (
    <div style={{ display: 'inline-block' }}>
      <Dialog
        title="标题"
        description="这是描述这是描述这是描述这是描述这是描述这是描述这是描述这是描述这是描述这是描述"
        visible={visible}
        centered
        destroyOnClose
        footer={() => null}
        onOk={() => setVisible(false)}
        onCancel={() => setVisible(false)}
      >
        <div>
          这是内容部分，这是内容部分这是内容部分这是内容部分这是内容部分这是内容部分这是内容部分这是内容部分这是内容部分这是内容部分这是内容部分
        </div>
      </Dialog>
      <Button onClick={() => setVisible(true)} type="primary">
        去除底部按钮
      </Button>
    </div>
  );
};

const RenderNoFooterCase2: FC<any> = () => {
  const [visible, setVisible] = React.useState<boolean>(false);
  return (
    <div style={{ display: 'inline-block' }}>
      <Dialog
        title="标题"
        description="这是描述这是描述这是描述这是描述这是描述这是描述这是描述这是描述这是描述这是描述"
        visible={visible}
        centered
        destroyOnClose
        footer={null}
        onOk={() => setVisible(false)}
        onCancel={() => setVisible(false)}
      >
        <div>
          这是内容部分，这是内容部分这是内容部分这是内容部分这是内容部分这是内容部分这是内容部分这是内容部分这是内容部分这是内容部分这是内容部分
        </div>
      </Dialog>
      <Button onClick={() => setVisible(true)} type="primary">
        去除底部按钮
      </Button>
    </div>
  );
};

const options = [111, 222, 333];

export const DifferentStates = () => (
  <div>
    <h2>Dialog</h2>
    <Divider text="基础" />
    <RenderBase
      okButtonProps={{ onClick: () => console.log('ok click') }}
      cancelButtonProps={{ onClick: () => console.log('cancel click') }}
    />
    <RenderBase text="隐藏右上角的叉" closable={false} />
    <RenderBase text="隐藏取消按钮" enableCancelButton={false} />
    <RenderBase text="不展示遮罩" mask={false} />
    <RenderBase text="遮罩不可点击" maskClosable={false} />
    <RenderBase text="取消ESC 关闭监听" keyboardClosable={false} />
    <RenderBase text="zIndex 1000" zIndex={1000} />
    <RenderBase
      text="safeMargin={{ top: 0, bottom: 0 }}"
      safeMargin={{ top: 0, bottom: 0 }}
      width="100%"
      style={{ top: 0, height: '100%' }}
      centered={false}
    />
    <RenderBase
      text="Context Menu 透传"
      onContextMenu={e => {
        e.preventDefault();
        const { clientX, clientY } = e;
        contextMenu.open({
          menu: (
            <Menu>
              {options.map(v => (
                <Menu.Item key={v}>{v}</Menu.Item>
              ))}
            </Menu>
          ),
          position: [clientX, clientY],
        });
      }}
    />

    {/* <RenderBase text="取消ESC 关闭监听" onAfterOpen={() => toast.success('onAfterOpen')} visible={true} /> */}
    <Divider text="自定义" />
    <RenderLoading />
    <RenderNoFooter />
    <RenderNoFooterCase2 />
    <h2>Confirm</h2>
    <Divider text="基础" />
    <Button
      onClick={() => {
        Dialog.confirm({
          title: '这是标题',
          content:
            '这是内容这是内容这是内容这是内容这是内容这是内容这是内容这是内容这是内容这是内容这是内容这是内容这是内容',
        });
      }}
    >
      base
    </Button>
    <Button
      onClick={() => {
        Dialog.success({
          title: '这是标题',
          content:
            '这是内容这是内容这是内容这是内容这是内容这是内容这是内容这是内容这是内容这是内容这是内容这是内容这是内容',
          maskClosable: true,
          enableCancelButton: true,
          okButtonProps: {
            onClick: () => console.log('ok click!'),
          },
          cancelButtonProps: {
            onClick: () => console.log('cancel click'),
          },
        });
      }}
    >
      success
    </Button>
    <Button
      danger
      onClick={() => {
        Dialog.error({
          title: '这是标题',
          content:
            '这是内容这是内容这是内容这是内容这是内容这是内容这是内容这是内容这是内容这是内容这是内容这是内容这是内容',
        });
      }}
    >
      error
    </Button>
    <Button
      onClick={() => {
        Dialog.info({
          title: '这是标题',
          content:
            '这是内容这是内容这是内容这是内容这是内容这是内容这是内容这是内容这是内容这是内容这是内容这是内容这是内容',
        });
      }}
    >
      info
    </Button>
    <Button
      onClick={() => {
        Dialog.warning({
          title: '这是标题',
          content:
            '这是内容这是内容这是内容这是内容这是内容这是内容这是内容这是内容这是内容这是内容这是内容这是内容这是内容',
        });
      }}
    >
      warning
    </Button>
    <Divider text="自定义" />
    <Button
      onClick={() => {
        Dialog.error({
          title: '确认删除吗？',
          okText: '删除',
          enableCancelButton: true,
          okButtonProps: {
            danger: true,
          },
        });
      }}
    >
      更改按钮
    </Button>
    <Button
      onClick={() => {
        Dialog.info({
          centered: false,
          bodyStyle: { top: '50px' },
          title: '自定义位置',
          content: '距离顶部 50 像素',
        });
      }}
    >
      自定义定位
    </Button>
    <Button
      onClick={() => {
        Dialog.error({
          icon: <BanOutlined />,
          centered: false,
          bodyStyle: { top: '50px' },
          title: '自定义位置',
          content: '距离顶部 50 像素',
          okButtonProps: {
            danger: true,
          },
        });
      }}
    >
      自定义 icon
    </Button>
    <Button
      onClick={() => {
        Dialog.error({
          icon: <BanOutlined />,
          centered: false,
          bodyStyle: { top: '50px' },
          title: '自定义位置',
          content: '距离顶部 50 像素',
          okButtonProps: {
            danger: true,
          },
          onOk: () =>
            new Promise((resolve, reject) => {
              setTimeout(Math.random() > 0.5 ? resolve : reject, 3000);
            }).catch(() => console.log('Oops errors!')),
        });
      }}
    >
      自定义关闭时间
    </Button>
    <Divider text="更新" />
    <Button
      onClick={() => {
        const { update } = Dialog.error({
          icon: <BanOutlined />,
          centered: false,
          bodyStyle: { top: '50px' },
          title: '自定义位置',
          content: '距离顶部 50 像素',
          okButtonProps: {
            danger: true,
          },
        });

        setTimeout(() => {
          update({
            okButtonProps: {
              loading: true,
            },
          });

          setTimeout(() => {
            update({
              okButtonProps: {
                loading: false,
              },
            });
          }, 2000);
        }, 2000);
      }}
    >
      打开2秒后更新按钮状态
    </Button>
    <Divider text="异步的回调" />
    <Button
      onClick={() => {
        Dialog.info({
          title: '123',
          onOk: onClose => {
            if (onClose) {
              onClose();
            }
          },
        });
      }}
    >
      ok
    </Button>
    <Button
      onClick={() => {
        Dialog.confirm({
          title: '123',
          onCancel: onClose => {
            if (onClose) {
              console.log('onClose');
              // onClose();
            }
          },
        });
      }}
    >
      点击取消后等待随机时间后关闭
    </Button>
    <Button
      onClick={() => {
        Dialog.confirm({
          title: '123',
          onCancel: () =>
            new Promise((resolve, reject) => {
              setTimeout(Math.random() > 0.5 ? resolve : reject, 3000);
            }).catch(() => console.log('Oops errors!')),
        });
      }}
    >
      ok
    </Button>
  </div>
);

const Context = React.createContext({ name: 'Default' });
export const useDialogDemo = () => {
  const isFirstMount = useFirstMountState();
  const [dialog, holder] = useDialog();
  const [flag, setFlag] = React.useState(true);
  const content = React.useMemo(
    () => (
      <>
        <Context.Consumer>{({ name }) => `Hello, ${name}!`}</Context.Consumer>
        <Button onClick={() => setFlag(s => !s)}>Change Context</Button>
      </>
    ),
    [],
  );

  if (isFirstMount) {
    const instance = dialog.info({
      content,
    });
    instance.update({
      title: 'Bamboo',
    });
  }

  const info = () => {
    const instance = dialog.info({
      content,
      onOk() {
        console.log('OK');
      },
      onCancel() {
        console.log('Cancel');
      },
    });

    instance.update({
      title: 'Bamboo',
    });
  };

  const success = () => {
    const instance = dialog.success({
      content,
      title: 'Do you want to delete these items?',
      onOk() {
        return new Promise((resolve, reject) => {
          setTimeout(Math.random() > 0.5 ? resolve : reject, 1000);
        }).catch(() => console.log('Oops errors!'));
      },
    });

    instance.update({
      title: 'Bamboo',
    });
  };

  return (
    <>
      <Context.Provider value={{ name: flag ? 'Universe Design' : 'Test Change' }}>
        <Space size="large">
          {holder}
          <Button onClick={info}>Info</Button>
          <Button onClick={success}>Success</Button>
          <Button
            onClick={() => {
              const instance = dialog.open({
                title: 'Something meaningless',
                children: content,
                // keyboardClosable: false,
                // maskClosable: false,
                onOk: () => {
                  // instance.destroy();
                },
                onCancel: (e, type) => {
                  console.log(e, type);
                  // instance.destroy();
                },
              });

              instance.update({
                title: 'Imperative Dialog',
              });
            }}
          >
            Modal
          </Button>
          <Button
            onClick={() => {
              const instance = dialog.confirm({
                content,
              });

              instance.destroy();
            }}
          >
            Destroy Immediately
          </Button>
        </Space>
      </Context.Provider>
    </>
  );
};

export const FormDemo = () => {
  const [visible, setVisible] = React.useState(false);
  const formRef = React.useRef(null);
  const showModal = () => setVisible(true);

  return (
    <>
      <Button type="primary" onClick={showModal}>
        Open Dialog
      </Button>
      <Dialog
        title="I am a title"
        description="This is a descriptive text with unlimited length and may not be available."
        visible={visible}
        onOk={e => {
          console.log('formRef: ', formRef.current);
          setVisible(false);
        }}
        destroyOnClose={false}
      >
        <Form ref={formRef} initialValues={{ a: 1 }}>
          <Form.Item name="a" label="label">
            <Radio.Group defaultValue={1} direction="vertical">
              <Radio value={1}>Radio content</Radio>
              <Radio value={2}>Radio content</Radio>
            </Radio.Group>
          </Form.Item>
        </Form>
      </Dialog>
    </>
  );
};

const selectOptions = new Array(20).fill('1').map(v => ({ value: '333' }));
const longContent = new Array(20).fill('1').map(x => <h1 key={x}>{x}</h1>);

export const RemoveScroll = () => {
  const [visible, setVisible] = React.useState(false);
  const [visible2, setVisible2] = React.useState(false);
  const getDialog = (props: { index: number }) => (
    <Dialog
      visible={visible}
      title="I am a title"
      description="This is a descriptive text with unlimited length and may not be available."
      onAfterClose={() => {
        console.log('onAfterClose');
      }}
      onOk={() => setVisible(false)}
      onAfterOpen={() => {
        console.log('onAfterOpen');
      }}
    >
      <Select options={selectOptions} />
      <Button onClick={() => setVisible2(true)}>Show Second Dialog </Button>

      <Dialog
        visible={visible2}
        onOk={() => setVisible2(false)}
        title="I am a title"
        description="This is a descriptive text with unlimited length and may not be available."
      />
    </Dialog>
  );

  return (
    <>
      <div
        style={{
          width: '100%',
          height: 10,
          marginBottom: 100,
          background: 'red',
        }}
      />
      <header
        style={{
          backgroundColor: 'rgba(0,0,0, .3)',
          position: 'fixed',
          top: 0,
          zIndex: 91000,
        }}
      >
        <Popover content={<div style={{ maxHeight: 200, overflow: 'scroll' }}>{longContent}</div>}>
          <Button>Popover </Button>
        </Popover>
        <Select options={selectOptions} />
      </header>
      {getDialog({ index: 1 })}

      <Button type="primary" onClick={() => setVisible(true)}>
        Open Dialog
      </Button>
      <div>{longContent}</div>
    </>
  );
};
export const Performance = () => {
  const [visible, setVisible] = useState(false);
  const [drawerVisible, setDrawerVisible] = useState(false);

  const showModal = () => setVisible(true);
  const content = [];
  for (let j = 0; j < 5; j++) {
    for (let i = 0; i < 10000; i++) {
      content.push(<div>我是一个div</div>);
    }
  }
  return (
    <>
      <p>initialFocusRef=null enableRestoreFocus=false 来关闭 focusLock, 以免页面节点数目过多时，性能过差</p>
      <Button type="primary" onClick={showModal}>
        Open Dialog
      </Button>
      <Button
        type="primary"
        onClick={e => {
          setDrawerVisible(true);
        }}
      >
        Open Drawer
      </Button>
      <Drawer
        initialFocusRef={null}
        enableRestoreFocus={false}
        title="Basic Drawer"
        visible={drawerVisible}
        onClose={e => {
          setDrawerVisible(false);
        }}
      >
        <p>Some contents...</p>
        <p>Some contents...</p>
        <p>Some contents...</p>
      </Drawer>
      <Dialog
        initialFocusRef={null}
        enableRestoreFocus={false}
        title="I am a title"
        description="This is a descriptive text with unlimited length and may not be available."
        visible={visible}
        onOk={e => {
          console.log(e);
          setVisible(false);
        }}
        onCancel={(e, cancelType) => {
          console.log(e);
          console.log(cancelType);
          setVisible(false);
        }}
        onAfterClose={() => {
          console.log('onAfterClose');
        }}
        onAfterOpen={() => {
          console.log('onAfterOpen');
        }}
      >
        <Radio.Group defaultValue={1} direction="vertical">
          <Radio value={1}>Radio content</Radio>
          <Radio value={2}>Radio content</Radio>
        </Radio.Group>
      </Dialog>
      {content}
    </>
  );
};
