// @universe-design/react/src/dialog/stories/screenshot.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import CConfirmContent from '../demo/confirm-content';
import CContentSize from '../demo/content-size';
import CFocuslockApi from '../demo/focuslock-api';
import CModalBody from '../demo/modal-body';
import CNewCase_0 from '../demo/new-case-0';
import CShadow from '../demo/shadow';
if (!(CConfirmContent as Story).storyName) {
  (CConfirmContent as Story).storyName = 'Confirm content';
}
if (!(CContentSize as Story).storyName) {
  (CContentSize as Story).storyName = 'Content size';
}
if (!(CFocuslockApi as Story).storyName) {
  (CFocuslockApi as Story).storyName = 'Focuslock api';
}
if (!(CModalBody as Story).storyName) {
  (CModalBody as Story).storyName = 'Modal body';
}
if (!(CNewCase_0 as Story).storyName) {
  (CNewCase_0 as Story).storyName = 'New case 0';
}
if (!(CShadow as Story).storyName) {
  (CShadow as Story).storyName = 'Shadow';
}

export default {
  title: 'Components/Dialog/截图测试',
  tags: ['autodocs'],
};

export { CConfirmContent, CContentSize, CFocuslockApi, CModalBody, CNewCase_0, CShadow };
