// @universe-design/react/src/dialog/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C12Size from '../demo/12.size';
import C14CuzHeaderFooter from '../demo/14.cuz-header-footer';
import C15DialogContent from '../demo/15.dialog-content';
import C16ConfirmContent from '../demo/16.confirm-content';
import C17UseDialog from '../demo/17.useDialog';
import C18LayoutMode from '../demo/18.layout-mode';
import CAsync from '../demo/async';
import CBasic from '../demo/basic';
import CConfirmRouter from '../demo/confirm-router';
import CConfirm from '../demo/confirm';
import CInfo from '../demo/info';
import CLocale from '../demo/locale';
import CManual from '../demo/manual';
import CPosition from '../demo/position';
import CWidth from '../demo/width';
if (!(C12Size as Story).storyName) {
  (C12Size as Story).storyName = '12. 对话框尺寸';
  (C12Size as Story).parameters = {
    docs: {
      description: {
        story:
          '使用`size`来设置对话框的尺寸。默认提供四种尺寸，`small`和`medium`为固定尺寸，`large`和`extra-large`为响应式。\n\n',
      },
    },
  };
}
if (!(C14CuzHeaderFooter as Story).storyName) {
  (C14CuzHeaderFooter as Story).storyName = '14. 自定义对话框头部和脚部';
  (C14CuzHeaderFooter as Story).parameters = {
    docs: {
      description: {
        story: '使用`header`和`footer`来设置对话框的头部和脚部。\n\n',
      },
    },
  };
}
if (!(C15DialogContent as Story).storyName) {
  (C15DialogContent as Story).storyName = '15. 直接使用对话框内容区域';
  (C15DialogContent as Story).parameters = {
    docs: {
      description: {
        story: '在一些场景下，也许你只希望使用对话框的样式而非弹出能力。\n\n',
      },
    },
  };
}
if (!(C16ConfirmContent as Story).storyName) {
  (C16ConfirmContent as Story).storyName = '16. 直接使用确认框内容区域';
  (C16ConfirmContent as Story).parameters = {
    docs: {
      description: {
        story: '在一些场景下，也许你只希望使用确认框的样式而非弹出能力。\n\n',
      },
    },
  };
}
if (!(C17UseDialog as Story).storyName) {
  (C17UseDialog as Story).storyName = '17. 能获取 Context 的命令式用法';
  (C17UseDialog as Story).parameters = {
    docs: {
      description: {
        story:
          '模块级别导出的 `Dialog.method()` API 使用独立的 React fiber root 进行渲染，无法跨 fiber tree 获取 context，这给实际使用带来了不便。`useDialog` 可以获取将 dialog 渲染到一个 `<ReactElements />` 节点中，这个 hook 返回两个值，第一个是 dialog 实例，提供调起命令，第二个参数是 `<ReactElements />` 节点，你需要手动把他放在可以获取到 context 的地方。\n\n',
      },
    },
  };
}
if (!(C18LayoutMode as Story).storyName) {
  (C18LayoutMode as Story).storyName = '18. 使用旧版布局';
  (C18LayoutMode as Story).parameters = {
    docs: {
      description: {
        story:
          "为了更好支持设计规范，我们在 `2.18.0` 修改了对话框的布局方式，由于修改布局方式，对于之前覆盖过对话框样式的用户可能产生 breaking change。默认情况下，我们会使用 `2.18.0` 新布局，如果用户侧存量代码较多，验证成本较高，可以使用 `<DialogLayoutMode mode='legacy' />` 修改子树中的对话框布局模式为 `legacy`。也可以通过 `DialogLayoutMode.setGlobal('legacy')` 全局设置，解决跨 fiberRoot 场景下无法共享 context 的问题。值的优先级顺序 **defaultValue < globalValue < contextValue**。\n\n建议渐进式验证并替换为新版 layout，`DialogLayoutMode` 将会在下个大版本中移除\n\n",
      },
    },
  };
}
if (!(CAsync as Story).storyName) {
  (CAsync as Story).storyName = '1. 异步关闭';
  (CAsync as Story).parameters = {
    docs: {
      description: {
        story: '点击确定后异步关闭对话框，例如提交表单。\n\n',
      },
    },
  };
}
if (!(CBasic as Story).storyName) {
  (CBasic as Story).storyName = '0. 基础用法';
  (CBasic as Story).parameters = {
    docs: {
      description: {
        story: '简单的用法。\n\n',
      },
    },
  };
}
if (!(CConfirmRouter as Story).storyName) {
  (CConfirmRouter as Story).storyName = '8. 销毁确认对话框';
  (CConfirmRouter as Story).parameters = {
    docs: {
      description: {
        story:
          '使用`Dialog.destroyAll()`可以销毁弹出的确认窗。通常用于路由监听当中，处理路由前进、后退不能销毁确认对话框的问题。\n\n',
      },
    },
  };
}
if (!(CConfirm as Story).storyName) {
  (CConfirm as Story).storyName = '3. 确认对话框';
  (CConfirm as Story).parameters = {
    docs: {
      description: {
        story: '使用`confirm()`可以快捷地弹出确认框。`onCancel`、`onOk`返回`Promise`可以延迟关闭。\n\n',
      },
    },
  };
}
if (!(CInfo as Story).storyName) {
  (CInfo as Story).storyName = '5. 信息提示';
  (CInfo as Story).parameters = {
    docs: {
      description: {
        story: '各种类型的信息提示，只提供一个按钮用于关闭。\n\n',
      },
    },
  };
}
if (!(CLocale as Story).storyName) {
  (CLocale as Story).storyName = '6. 自定义按钮文字';
  (CLocale as Story).parameters = {
    docs: {
      description: {
        story: '设置`okText`与`cancelText`可以自定义按钮文字。\n\n',
      },
    },
  };
}
if (!(CManual as Story).storyName) {
  (CManual as Story).storyName = '7. 手动更新和移除';
  (CManual as Story).parameters = {
    docs: {
      description: {
        story: '手动更新和关闭`Dialog.method`方式创建的对话框。\n\n',
      },
    },
  };
}
if (!(CPosition as Story).storyName) {
  (CPosition as Story).storyName = '7. 自定义位置';
  (CPosition as Story).parameters = {
    docs: {
      description: {
        story: '通过`centered`或`style`来设置对话框位置。默认为垂直居中布局，`centered`模式下`style.top`会被重置。\n\n',
      },
    },
  };
}
if (!(CWidth as Story).storyName) {
  (CWidth as Story).storyName = '13. 自定义对话框的宽度';
  (CWidth as Story).parameters = {
    docs: {
      description: {
        story: '使用`width`来设置对话框的宽度。该属性可以在无法使用`size`的场景下强行指定对话框宽度。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Dialog/官网示例',
  tags: ['autodocs'],
};

export {
  C12Size,
  C14CuzHeaderFooter,
  C15DialogContent,
  C16ConfirmContent,
  C17UseDialog,
  C18LayoutMode,
  CAsync,
  CBasic,
  CConfirmRouter,
  CConfirm,
  CInfo,
  CLocale,
  CManual,
  CPosition,
  CWidth,
};
