// @byted-service/env/lib/idc_util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IDC = void 0;
exports.IDC = {
    UnknownIDC: '-',
    DC_BR: 'br',
    DC_JP: 'jp',
    DC_IN: 'in',
    DC_FRAWS: 'fraws',
    DC_MAWSJP: 'mawsjp',
    DC_AGCQ: 'agcq',
    DC_AGGDSZ: 'aggdsz',
    DC_AGHBWH: 'aghbwh',
    DC_AGJSNJ: 'agjsnj',
    DC_AGLNSY: 'aglnsy',
    DC_AGSDQD: 'agsdqd',
    DC_AGSXLQ: 'agsxlq',
    DC_AGSXXA: 'agsxxa',
    DC_AGSY: 'agsy',
    DC_ALIEC1: 'aliec1',
    DC_ALIEC2: 'aliec2',
    DC_ALIGDSZ: 'aligdsz',
    DC_ALINC2: 'alinc2',
    DC_ALISC1: 'alisc1',
    DC_ALISG: 'alisg',
    DC_ALISH: 'alish',
    DC_ALIVA: 'aliva',
    DC_AWSBH: 'awsbh',
    DC_AWSBR: 'awsbr',
    DC_AWSFR: 'awsfr',
    DC_AWSIN: 'awsin',
    DC_AWSJP: 'awsjp',
    DC_AWSJPGM: 'awsjpgm',
    DC_AWSNC1: 'awsnc1',
    DC_AWSNWC1: 'awsnwc1',
    DC_AWSSG: 'awssg',
    DC_AWSSGGM: 'awssggm',
    DC_AWSVAC: 'awsvac',
    DC_AWSVAGM: 'awsvagm',
    DC_BJGS: 'bjgs',
    DC_BJLGY: 'bjlgy',
    DC_BOE: 'boe',
    DC_BOE2: 'boe2',
    DC_IBOE: 'boei18n',
    DC_CA: 'ca',
    DC_COF: 'cof',
    DC_DEVBOX: 'devbox',
    DC_DEVBOXI18N: 'devboxi18n',
    DC_GALINC2: 'galinc2',
    DC_GALISG: 'galisg',
    DC_GALIVA: 'galiva',
    DC_GCPAU: 'gcpau',
    DC_GCPBE: 'gcpbe',
    DC_GCPBR: 'gcpbr',
    DC_GCPCA: 'gcpca',
    DC_GCPCH: 'gcpch',
    DC_GCPDE: 'gcpde',
    DC_GCPFI: 'gcpfi',
    DC_GCPGB: 'gcpgb',
    DC_GCPHK: 'gcphk',
    DC_GCPIN: 'gcpin',
    DC_GCPJPOSA: 'gcpjposa',
    DC_GCPJPTKY: 'gcpjptky',
    DC_GCPNL: 'gcpnl',
    DC_GCPSG: 'gcpsg',
    DC_GCPTW: 'gcptw',
    DC_GCPUSCBF: 'gcpuscbf',
    DC_GCPUSIAD: 'gcpusiad',
    DC_GZTOBIAAS: 'gztobiaas',
    DC_HKCJ: 'hkcj',
    DC_HL: 'hl',
    DC_HLTOBIAAS: 'hltobiaas',
    DC_NTTOBIAAS: 'nttobiaas',
    DC_HY: 'hy',
    DC_IND: 'ind',
    DC_INDDP: 'inddp',
    DC_JPSAAS: 'jpsaas',
    DC_KSRU: 'ksru',
    DC_LF: 'lf',
    DC_LFTOBIAAS: 'lftobiaas',
    DC_LQ: 'lq',
    DC_MALIVA: 'maliva',
    DC_QDTOB: 'qdtob',
    DC_QDTOBIAAS: 'qdtobiaas',
    DC_SDQDCJ: 'sdqdcj',
    DC_SDQD: 'sdqd',
    DC_SG: 'sg',
    DC_SG1: 'sg1',
    DC_SG2: 'sg2',
    DC_SGCOMM1: 'sgcomm1',
    DC_SGCOMPLIANCE: 'sgcompliance',
    DC_SGDT: 'sgdt',
    DC_SGEA1: 'sgea1',
    DC_SGEA2: 'sgea2',
    DC_SGEE1: 'sgee1',
    DC_SGEE2: 'sgee2',
    DC_SGEE3: 'sgee3',
    DC_SGSAAS1LARKIDC1: 'sgsaas1larkidc1',
    DC_SGSAAS1LARKIDC2: 'sgsaas1larkidc2',
    DC_SGSAAS1LARKIDC3: 'sgsaas1larkidc3',
    DC_SHTOBIAAS: 'shtobiaas',
    DC_SYKA2LARK: 'syka2lark',
    DC_SYKA3LARK: 'syka3lark',
    DC_SYKALARK: 'sykalark',
    DC_USEAST2A: 'useast2a',
    DC_USEAST3: 'useast3',
    DC_USEAST4: 'useast4',
    DC_USEAST5: 'useast5',
    DC_USEAST8: 'useast8',
    DC_USEASTDT: 'useastdt',
    DC_USWEST1A: 'uswest1a',
    DC_USTSENTRY: 'ustsentry',
    DC_USORDX: 'usordx',
    DC_VA: 'va',
    DC_WJ: 'wj',
    DC_YCRU: 'ycru',
    DC_YG: 'yg',
    DC_VOLCNT: 'volcnt',
    DC_VOLCGZ: 'volcgz',
    DC_FRACOMPLIANCE: 'fracompliance',
    DC_BOETTP: 'boettp',
    DC_MY: 'my',
    DC_PD: 'pd',
    DC_IE: 'ie',
    DC_LFZG: 'lfzg',
    DC_HLZG: 'hlzg',
    DC_XFLONLINE: 'xflonline',
    DC_SINFBOE: 'sinfboe',
    DC_SINFONLINE: 'sinfonline',
    DC_SINFONLINEA: 'sinfonlinea',
    DC_SINFONLINEC: 'sinfonlinec',
    DC_SINFSHA: 'sinfsha',
    DC_IE2: 'ie2',
    DC_DE: 'de',
    DC_ID1A: 'id1a',
    DC_ID2A: 'id2a',
    DC_USEAST11A: 'useast11a',
    DC_IEDT: 'iedt',
    DC_DEDT: 'dedt',
    DC_NO1A: 'no1a',
    DC_MYA: 'mya',
    DC_MYB: 'myb',
    DC_HJ: 'hj',
    DC_HJZG: 'hjzg',
    DC_USEAST9A: 'useast9a',
    DC_MY2: 'my2',
    DC_MY3: 'my3',
    DC_MYCISB: 'mycisb',
    DC_SGCISA: 'sgcisa',
    DC_USEAST10A: 'useast10a',
    DC_MY4A: 'my4a',
    DC_USEAST2B: 'useast2b', // Emerald/OCI-Emerald/OCI-New, US-EastRed 新机房
};
