// @byted-service/env/lib/ip.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isIPv6Only = exports.isIPv4Only = exports.isDualStack = exports.hasIPv6 = exports.hasIPv4 = exports.getIPStr = exports.getIP = exports.getIPv6 = exports.getIPv4 = exports.getHostIPV6 = exports.getHostIPV4 = exports.getHostIP = exports.init = void 0;
const tslib_1 = require("tslib");
const os_1 = tslib_1.__importDefault(require("os"));
const net_1 = tslib_1.__importDefault(require("net"));
const tce_1 = require("./tce");
let tceIPV4 = '';
let tceIPV6 = '';
let hostIPV4 = '';
let hostIPV6 = '';
const init = () => {
    tceIPV4 = '';
    tceIPV6 = '';
    hostIPV4 = '';
    hostIPV6 = '';
    if (process.env.IS_TCE_DOCKER_ENV === '1') {
        tceIPV4 = process.env.MY_HOST_IP || '';
        tceIPV6 = process.env.MY_HOST_IPV6 || '';
    }
    else {
        const ifs = os_1.default.networkInterfaces();
        const keys = Object.keys(ifs);
        for (let i = 0; i < keys.length; i++) {
            for (const item of ifs[keys[i]] || []) {
                if (!item.internal && net_1.default.isIPv4(item.address) && !hostIPV4) {
                    hostIPV4 = item.address;
                }
                if (!item.internal && net_1.default.isIPv6(item.address) && !hostIPV6) {
                    hostIPV6 = item.address;
                }
                if (hostIPV4 && hostIPV6) {
                    break;
                }
            }
            if (hostIPV4 && hostIPV6) {
                break;
            }
        }
    }
};
exports.init = init;
/**
 * HostIP 保留接口，为了兼容性优先返回 IPV4，其次是 IPV6
 */
const getHostIP = () => {
    return (0, exports.getHostIPV4)() || (0, exports.getHostIPV6)();
};
exports.getHostIP = getHostIP;
/**
 * HostIPV4 返回 ipv4 的 string
 */
const getHostIPV4 = () => (0, exports.getIPv4)();
exports.getHostIPV4 = getHostIPV4;
/**
 * HostIPV6 返回 ipv6 的 string
 */
const getHostIPV6 = () => (0, exports.getIPv6)();
exports.getHostIPV6 = getHostIPV6;
/**
 * IPV4 返回 ipv4
 */
const getIPv4 = () => {
    if ((0, tce_1.isInTCE)() && tceIPV4) {
        return tceIPV4;
    }
    return hostIPV4;
};
exports.getIPv4 = getIPv4;
/**
 * IPV6 返回 ipv6
 */
const getIPv6 = () => {
    if ((0, tce_1.isInTCE)() && tceIPV6) {
        return tceIPV6;
    }
    return hostIPV6;
};
exports.getIPv6 = getIPv6;
/**
 * IP 优先返回 IPV6，如果没有 IPV6 则返回 IPV4
 * 适用于 不在意获取到的是 IPV4 还是 IPV6 的业务
 */
const getIP = () => {
    if ((0, exports.getIPv6)()) {
        return (0, exports.getIPv6)();
    }
    return (0, exports.getIPv4)();
};
exports.getIP = getIP;
/**
 * IPStr 同上，返回 string
 */
const getIPStr = () => (0, exports.getIP)();
exports.getIPStr = getIPStr;
/**
 * HasIPV4
 */
const hasIPv4 = () => {
    return ((0, tce_1.isInTCE)() && tceIPV4 !== '') || hostIPV4 !== '';
};
exports.hasIPv4 = hasIPv4;
/**
 * HasIPV6
 */
const hasIPv6 = () => {
    return ((0, tce_1.isInTCE)() && tceIPV6 !== '') || hostIPV6 !== '';
};
exports.hasIPv6 = hasIPv6;
/**
 * IsDualStack
 */
const isDualStack = () => {
    return (0, exports.hasIPv4)() && (0, exports.hasIPv6)();
};
exports.isDualStack = isDualStack;
/**
 * IsIPV4Only
 */
const isIPv4Only = () => {
    return (0, exports.hasIPv4)() && !(0, exports.hasIPv6)();
};
exports.isIPv4Only = isIPv4Only;
/**
 * IsIPV6Only
 */
const isIPv6Only = () => {
    return !(0, exports.hasIPv4)() && (0, exports.hasIPv6)();
};
exports.isIPv6Only = isIPv6Only;
