// @byted-service/env/lib/nodejs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isTest = exports.isProd = exports.init = exports.defaultCloudEnv = exports.defaultImageVersion = exports.defaultPodName = exports.defaultStage = exports.defaultEnv = exports.defaultIDC = exports.defaultCluster = exports.defaultPSM = void 0;
const product_1 = require("./product");
exports.defaultPSM = '-';
exports.defaultCluster = 'default';
exports.defaultIDC = '';
exports.defaultEnv = 'prod';
exports.defaultStage = '-';
exports.defaultPodName = '-';
exports.defaultImageVersion = '-';
exports.defaultCloudEnv = '-';
let env = '';
const init = () => {
    env = '';
};
exports.init = init;
/**
 * 是否是 prod 环境，相比于 env.isProduct() 多支持了 NODE_ENV 环境变量
 */
const isProd = () => {
    if (env) {
        return env === 'prod';
    }
    if ((0, product_1.isTesting)() || (0, product_1.isBoe)()) {
        env = 'test';
        return false;
    }
    if (process.env.NODE_ENV === 'production' ||
        process.env.NODE_ENV === 'prod') {
        env = 'prod';
        return true;
    }
    if ((0, product_1.isProduct)()) {
        env = 'prod';
        return true;
    }
    return false;
};
exports.isProd = isProd;
/**
 * 是否是 testing or boe 环境
 */
const isTest = () => {
    if (env) {
        return env === 'test';
    }
    if ((0, product_1.isTesting)() || (0, product_1.isBoe)()) {
        env = 'test';
        return true;
    }
    return false;
};
exports.isTest = isTest;
