// @byted-service/env/lib/service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setCluster = exports.getCluster = exports.setPSM = exports.getPSM = exports.init = exports.ClusterDefault = exports.PSMUnknown = void 0;
exports.PSMUnknown = '-';
exports.ClusterDefault = 'default';
let psm = '';
let cluster = '';
const init = () => {
    psm = '';
    cluster = '';
    psm = process.env.LOAD_SERVICE_PSM || '';
    if (!psm) {
        psm = process.env.PSM || '';
    }
    if (!psm) {
        psm = process.env.TCE_PSM || '';
    }
    if (!psm) {
        psm = exports.PSMUnknown;
    }
    cluster = process.env.SERVICE_CLUSTER || '';
    if (!cluster) {
        cluster = exports.ClusterDefault;
    }
};
exports.init = init;
/**
 * PSM .
 */
const getPSM = () => {
    return psm;
};
exports.getPSM = getPSM;
/**
 * SetPSM is used for unit test.
 */
const setPSM = (psm_) => {
    psm = psm_;
};
exports.setPSM = setPSM;
/**
 * Cluster .
 */
const getCluster = () => {
    return cluster;
};
exports.getCluster = getCluster;
/**
 * SetCluster is used for unit test.
 */
const setCluster = (cluster_) => {
    cluster = cluster_;
};
exports.setCluster = setCluster;
