// @byted-service/env/lib/tce.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isHostNetwork = exports.inVolcanoCloud = exports.getCloudEnv = exports.getImageVersion = exports.getPodName = exports.isInTCE = exports.init = void 0;
let inTCE = false;
let podName = '';
let imageVersion = '';
let cloudEnv = '';
let _inVolcanoCloud = false;
let _isHostNetwork = false;
const init = () => {
    inTCE = false;
    podName = '';
    imageVersion = '';
    cloudEnv = '';
    _inVolcanoCloud = false;
    _isHostNetwork = false;
    if (process.env.IS_TCE_DOCKER_ENV === '1') {
        inTCE = true;
    }
    podName = process.env.MY_POD_NAME || '';
    if (!podName) {
        podName = '-';
    }
    imageVersion = process.env.CURRENT_VERSION || '';
    if (!imageVersion) {
        imageVersion = '-';
    }
    cloudEnv = process.env.PAAS_CLOUD_ENV || '';
    if (!cloudEnv) {
        cloudEnv = '-';
    }
    if (cloudEnv === 'VOLCANO') {
        _inVolcanoCloud = true;
    }
    const ihn = process.env.IS_HOST_NETWORK || '';
    if (ihn === '1') {
        _isHostNetwork = true;
    }
};
exports.init = init;
/**
 * InTCE
 */
const isInTCE = () => {
    return inTCE;
};
exports.isInTCE = isInTCE;
/**
 * PodName
 */
const getPodName = () => {
    return podName;
};
exports.getPodName = getPodName;
/**
 * ImageVersion
 */
const getImageVersion = () => {
    return imageVersion;
};
exports.getImageVersion = getImageVersion;
/**
 * CloudEnv
 */
const getCloudEnv = () => {
    return cloudEnv;
};
exports.getCloudEnv = getCloudEnv;
/**
 * InVolcanoCloud
 */
const inVolcanoCloud = () => {
    return _inVolcanoCloud;
};
exports.inVolcanoCloud = inVolcanoCloud;
/**
 * IsHostNetwork
 * 网络模式是否为 host 或 autohost 模式
 */
const isHostNetwork = () => {
    return _isHostNetwork;
};
exports.isHostNetwork = isHostNetwork;
