// @byted-service/env/lib/vregion.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tt = exports.getGovernanceVRegionFromVDC = exports.getVRegionFromHost = exports.getCurrentVRegion = exports.getVRegionFromVDC = exports.VREGION = exports.VDC = void 0;
const idc_1 = require("./idc");
const vregion_data_1 = require("./internal/vregion.data");
Object.defineProperty(exports, "VDC", { enumerable: true, get: function () { return vregion_data_1.VDC; } });
Object.defineProperty(exports, "VREGION", { enumerable: true, get: function () { return vregion_data_1.VREGION; } });
const vdcToVregionMap = vregion_data_1.VDC_TO_VREGION_MAP;
/**
 * GetVRegionFromVDC gets VRegion that VDC belongs to
 */
const getVRegionFromVDC = (vdc) => {
    const vRegion = vdcToVregionMap[vdc];
    if (vRegion) {
        return vRegion;
    }
    return vregion_data_1.VREGION.VREGION_UNKNOWN;
};
exports.getVRegionFromVDC = getVRegionFromVDC;
/**
 * GetCurrentVRegion locates current service's VRegion
 */
const getCurrentVRegion = () => {
    return (0, exports.getVRegionFromVDC)((0, idc_1.getIDC)());
};
exports.getCurrentVRegion = getCurrentVRegion;
/**
 * GetVRegionFromHost gets VRegion that host belongs to
 */
const getVRegionFromHost = (ip) => {
    return (0, exports.getVRegionFromVDC)((0, idc_1.getIDCFromHost)(ip));
};
exports.getVRegionFromHost = getVRegionFromHost;
/**
 * vRegion->vRegion in governance special case
 */
const vRegionToGovernanceRegionMap = {
    [vregion_data_1.VREGION.VREGION_CHINAFINTECH]: vregion_data_1.VREGION.VREGION_CHINANORTH,
    [vregion_data_1.VREGION.VREGION_CHINAAGGREGATION]: vregion_data_1.VREGION.VREGION_CHINANORTH,
    [vregion_data_1.VREGION.VREGION_EUROPECENTRAL]: vregion_data_1.VREGION.VREGION_SINGAPORECENTRAL,
};
/**
 * GetGovernanceVRegionFromVDC Get the region that real governace rules store
 */
const getGovernanceVRegionFromVDC = (vdc) => {
    const vRegion = (0, exports.getVRegionFromVDC)(vdc);
    const ret = vRegionToGovernanceRegionMap[vRegion];
    if (ret) {
        return ret;
    }
    return vRegion;
};
exports.getGovernanceVRegionFromVDC = getGovernanceVRegionFromVDC;
//////////////////////////////////////////////////////
/**
 * Helper functions for TikTok bussiness only
 * Sync with tiktok/region_lib
 */
function pick(obj, keys) {
    const result = {};
    for (const key of keys) {
        if (key in obj) {
            result[key] = obj[key];
        }
    }
    return result;
}
/**
 * vdc for TT
 */
const TT_VDC = Object.assign({
    VDC_UNKNOWN: '-',
    VDC_NO: 'no',
}, pick(vregion_data_1.VDC, [
    'VDC_MALIVA',
    'VDC_USEAST10A',
    'VDC_SG1',
    'VDC_MY',
    'VDC_MY2',
    'VDC_MY3',
    'VDC_USEAST2A',
    'VDC_USEAST2B',
    'VDC_USEAST5',
    'VDC_USEAST8',
    'VDC_IE',
    'VDC_BOEI18N',
    'VDC_BOETTP',
    'VDC_NO1A',
]));
/**
 * vRegion for TT
 */
const TT_VREGION = pick(vregion_data_1.VREGION, [
    'VREGION_UNKNOWN',
    'VREGION_SINGAPORECENTRAL',
    'VREGION_USEAST',
    'VREGION_USEASTRED',
    'VREGION_USTTP',
    'VREGION_USTTP2',
    'VREGION_EUTTP',
    'VREGION_EUTTP2',
]);
/**
 * vGEO for TT
 */
const TT_VGEO = {
    VGEO_UNKNOWN: '-',
    VGEO_US: 'VGeo-US',
    VGEO_EU: 'VGeo-EU',
    VGEO_ROW: 'VGeo-ROW',
};
const TT_COMPLIANCE_REGION = {
    CR_UNKNOWN: '-',
    CR_ROW: 'ROW',
    CR_TEXAS: 'Texas',
    CR_CLOVER: 'Clover',
};
/**
 * TT vDC -> TT vRegion
 */
const ttVDCToVRegionMap = {
    [TT_VDC.VDC_UNKNOWN]: TT_VREGION.VREGION_UNKNOWN,
    [TT_VDC.VDC_MALIVA]: TT_VREGION.VREGION_USEAST,
    [TT_VDC.VDC_USEAST10A]: TT_VREGION.VREGION_USEAST,
    [TT_VDC.VDC_SG1]: TT_VREGION.VREGION_SINGAPORECENTRAL,
    [TT_VDC.VDC_MY]: TT_VREGION.VREGION_SINGAPORECENTRAL,
    [TT_VDC.VDC_MY2]: TT_VREGION.VREGION_SINGAPORECENTRAL,
    [TT_VDC.VDC_MY3]: TT_VREGION.VREGION_SINGAPORECENTRAL,
    [TT_VDC.VDC_USEAST2A]: TT_VREGION.VREGION_USEASTRED,
    [TT_VDC.VDC_USEAST2B]: TT_VREGION.VREGION_USEASTRED,
    [TT_VDC.VDC_USEAST5]: TT_VREGION.VREGION_USTTP,
    [TT_VDC.VDC_USEAST8]: TT_VREGION.VREGION_USTTP2,
    [TT_VDC.VDC_IE]: TT_VREGION.VREGION_EUTTP,
    [TT_VDC.VDC_NO]: TT_VREGION.VREGION_EUTTP2,
    [TT_VDC.VDC_NO1A]: TT_VREGION.VREGION_EUTTP2,
};
/**
 * TT vRegion -> TT vgeo
 */
const vRegionToVGeoMap = {
    [TT_VREGION.VREGION_UNKNOWN]: TT_VGEO.VGEO_UNKNOWN,
    [TT_VREGION.VREGION_USEAST]: TT_VGEO.VGEO_ROW,
    [TT_VREGION.VREGION_SINGAPORECENTRAL]: TT_VGEO.VGEO_ROW,
    [TT_VREGION.VREGION_USEASTRED]: TT_VGEO.VGEO_EU,
    [TT_VREGION.VREGION_USTTP]: TT_VGEO.VGEO_US,
    [TT_VREGION.VREGION_USTTP2]: TT_VGEO.VGEO_US,
    [TT_VREGION.VREGION_EUTTP]: TT_VGEO.VGEO_EU,
    [TT_VREGION.VREGION_EUTTP2]: TT_VGEO.VGEO_EU,
};
/**
 * getVDC, only return known VDC for TT
 */
function getVDC() {
    const idc = (0, idc_1.getIDC)();
    return ttVDCToVRegionMap[idc] ? idc : TT_VDC.VDC_UNKNOWN;
}
function getVRegion() {
    return ttVDCToVRegionMap[getVDC()] || TT_VREGION.VREGION_UNKNOWN;
}
function getVGeo() {
    if ((0, idc_1.getIDC)() === vregion_data_1.VDC.VDC_BOETTP) {
        return TT_VGEO.VGEO_US;
    }
    return vRegionToVGeoMap[getVRegion()] || TT_VREGION.VREGION_UNKNOWN;
}
function ttGetVRegionFromVDC(vdc) {
    return ttVDCToVRegionMap[vdc] || TT_VREGION.VREGION_UNKNOWN;
}
function getVGeoFromVDC(vdc) {
    return vRegionToVGeoMap[ttGetVRegionFromVDC(vdc)] || TT_VGEO.VGEO_UNKNOWN;
}
function getVGeoFromVRegion(vRegion) {
    return vRegionToVGeoMap[vRegion] || TT_VGEO.VGEO_UNKNOWN;
}
function isVGeoUS() {
    return getVGeo() === TT_VGEO.VGEO_US;
}
function isVGeoEU() {
    return getVGeo() === TT_VGEO.VGEO_EU;
}
const regionCodeTexasList = new Set(['us', 'as', 'gu', 'mp', 'pr', 'vi', 'um']);
const regionCodeList = new Set([
    'ad',
    'ae',
    'af',
    'ag',
    'ai',
    'al',
    'am',
    'ao',
    'aq',
    'ar',
    'as',
    'at',
    'au',
    'aw',
    'ax',
    'az',
    'ba',
    'bb',
    'bd',
    'be',
    'bf',
    'bg',
    'bh',
    'bi',
    'bj',
    'bl',
    'bm',
    'bn',
    'bo',
    'bq',
    'br',
    'bs',
    'bt',
    'bv',
    'bw',
    'by',
    'bz',
    'ca',
    'cc',
    'cd',
    'cf',
    'cg',
    'ch',
    'ci',
    'ck',
    'cl',
    'cm',
    'cn',
    'co',
    'cr',
    'cu',
    'cv',
    'cw',
    'cx',
    'cy',
    'cz',
    'de',
    'dj',
    'dk',
    'dm',
    'do',
    'dz',
    'ec',
    'ee',
    'eg',
    'eh',
    'er',
    'es',
    'et',
    'fi',
    'fj',
    'fk',
    'fm',
    'fo',
    'fr',
    'ga',
    'gb',
    'gd',
    'ge',
    'gf',
    'gg',
    'gh',
    'gi',
    'gl',
    'gm',
    'gn',
    'gp',
    'gq',
    'gr',
    'gs',
    'gt',
    'gu',
    'gw',
    'gy',
    'hk',
    'hm',
    'hn',
    'hr',
    'ht',
    'hu',
    'id',
    'ie',
    'il',
    'im',
    'in',
    'io',
    'iq',
    'ir',
    'is',
    'it',
    'je',
    'jm',
    'jo',
    'jp',
    'ke',
    'kg',
    'kh',
    'ki',
    'km',
    'kn',
    'kp',
    'kr',
    'kw',
    'ky',
    'kz',
    'la',
    'lb',
    'lc',
    'li',
    'lk',
    'lr',
    'ls',
    'lt',
    'lu',
    'lv',
    'ly',
    'ma',
    'mc',
    'md',
    'me',
    'mf',
    'mg',
    'mh',
    'mk',
    'ml',
    'mm',
    'mn',
    'mo',
    'mp',
    'mq',
    'mr',
    'ms',
    'mt',
    'mu',
    'mv',
    'mw',
    'mx',
    'my',
    'mz',
    'na',
    'nc',
    'ne',
    'nf',
    'ng',
    'ni',
    'nl',
    'no',
    'np',
    'nr',
    'nu',
    'nz',
    'om',
    'pa',
    'pe',
    'pf',
    'pg',
    'ph',
    'pk',
    'pl',
    'pm',
    'pn',
    'pr',
    'ps',
    'pt',
    'pw',
    'py',
    'qa',
    're',
    'ro',
    'rs',
    'ru',
    'rw',
    'sa',
    'sb',
    'sc',
    'sd',
    'se',
    'sg',
    'sh',
    'si',
    'sj',
    'sk',
    'sl',
    'sm',
    'sn',
    'so',
    'sr',
    'ss',
    'st',
    'sv',
    'sx',
    'sy',
    'sz',
    'tc',
    'td',
    'tf',
    'tg',
    'th',
    'tj',
    'tk',
    'tl',
    'tm',
    'tn',
    'to',
    'tr',
    'tt',
    'tv',
    'tw',
    'tz',
    'ua',
    'ug',
    'um',
    'us',
    'uy',
    'uz',
    'va',
    'vc',
    've',
    'vg',
    'vi',
    'vn',
    'vu',
    'wf',
    'ws',
    'xk',
    'ye',
    'yt',
    'za',
    'zm',
    'zw',
]);
const regionCodeCloverList = new Set([
    'at',
    'be',
    'bg',
    'ch',
    'cy',
    'cz',
    'de',
    'dk',
    'ee',
    'es',
    'fi',
    'fr',
    'gb',
    'gf',
    'gp',
    'gr',
    'hr',
    'hu',
    'ie',
    'is',
    'it',
    'li',
    'lt',
    'lu',
    'lv',
    'mf',
    'mq',
    'mt',
    'nl',
    'no',
    'pl',
    'pt',
    're',
    'ro',
    'se',
    'si',
    'sk',
    'yt',
]);
function isTexasRegion(regionCode) {
    return regionCodeTexasList.has(regionCode.toLowerCase());
}
function isCloverRegion(regionCode) {
    return regionCodeCloverList.has(regionCode.toLowerCase());
}
function getCloverRegionList() {
    return Array.from(regionCodeCloverList);
}
function isROWRegion(regionCode) {
    if (isTexasRegion(regionCode) || isCloverRegion(regionCode)) {
        return false;
    }
    return regionCodeList.has(regionCode.toLowerCase());
}
function getComplianceRegion(regionCode) {
    if (isTexasRegion(regionCode)) {
        return TT_COMPLIANCE_REGION.CR_TEXAS;
    }
    if (isCloverRegion(regionCode)) {
        return TT_COMPLIANCE_REGION.CR_CLOVER;
    }
    if (isROWRegion(regionCode)) {
        return TT_COMPLIANCE_REGION.CR_ROW;
    }
    return TT_COMPLIANCE_REGION.CR_UNKNOWN;
}
exports.tt = {
    VDC: TT_VDC,
    VREGION: TT_VREGION,
    VGEO: TT_VGEO,
    getVDC,
    getVRegion,
    getVRegionFromVDC: ttGetVRegionFromVDC,
    getVGeo,
    getVGeoFromVDC,
    getVGeoFromVRegion,
    isVGeoUS,
    isVGeoEU,
    isTexasRegion,
    isCloverRegion,
    getCloverRegionList,
    isROWRegion,
    getComplianceRegion,
    COMPLIANCE_REGION: TT_COMPLIANCE_REGION,
    vRegionToVGeoMap,
    vdcToVregionMap: ttVDCToVRegionMap,
};
