// @universe-design/styles/cjs/ccm-ccm-mindnote/index.light.atom-raw.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ccmtoken_mindnote_highlightcolor_yellow_raw = exports.ccmtoken_mindnote_highlightcolor_turquoise_raw = exports.ccmtoken_mindnote_highlightcolor_red_raw = exports.ccmtoken_mindnote_highlightcolor_neutral_raw = exports.ccmtoken_mindnote_highlightcolor_lime_raw = exports.ccmtoken_mindnote_highlightcolor_carmine_raw = exports.ccmtoken_mindnote_highlightcolor_blue_raw = void 0;
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
exports.ccmtoken_mindnote_highlightcolor_blue_raw = '194, 212, 255';
exports.ccmtoken_mindnote_highlightcolor_carmine_raw = '254, 226, 242';
exports.ccmtoken_mindnote_highlightcolor_lime_raw = '227, 240, 163';
exports.ccmtoken_mindnote_highlightcolor_neutral_raw = '222, 224, 227';
exports.ccmtoken_mindnote_highlightcolor_red_raw = '253, 198, 196';
exports.ccmtoken_mindnote_highlightcolor_turquoise_raw = '196, 242, 236';
exports.ccmtoken_mindnote_highlightcolor_yellow_raw = '250, 237, 194';
