// @universe-design/react/es/dialog/Confirm.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.ugr6219{margin-top:auto;}
import uwejdxd from "./../.stylex/uwejdxd";
//.u11qmguv{margin-right:0;}
import u1purjcd from "./../.stylex/u1purjcd";
//.u10jk5vf{margin-bottom:auto;}
import uuhcnxi from "./../.stylex/uuhcnxi";
//.u1tyq0we{margin-left:0;}
import u2o5q4q from "./../.stylex/u2o5q4q";
import React, { useEffect, useState } from 'react';
import ErrorFilled from '@universe-design/icons-react/ErrorFilled';
import InfoFilled from '@universe-design/icons-react/InfoFilled';
import SucceedFilled from '@universe-design/icons-react/SucceedFilled';
import WarningFilled from '@universe-design/icons-react/WarningFilled';
import clsx from 'clsx';
import stylex from '@universe-design/stylex';
import { cssTypographyBody0, cssTypographyTitle4 } from '../stylex';
import { sxDialogIconWrapper, sxDialogIcon, sxModalContent } from './sx';
import { Dialog } from './dialog';
import { Button } from '../button';
import { useCssScope, useLocale } from '../config-provider/hooks';
import { isFunction, isNonNullable } from '../util';
import { useEffectOnce } from '../hooks';
import { FirstLineAlign } from '../shared/components/FirstLineAlign';
import './style';

/**
 * @public
 * @docgen
 */

export const ConfirmContent = props => {
  const {
    icon: iconProp,
    title,
    content,
    cancelText,
    okText,
    cancelButtonProps,
    okButtonProps,
    enableCancelButton = true,
    autoFocusButton = 'ok',
    bodyStyle,
    onCancel,
    onOk,
    onClose,
    type: typeProp,
    extraAction,
    // style
    style,
    className,
    // ref
    okButtonRef,
    cancelButtonRef,
    sx = []
  } = props;
  const type = typeProp !== null && typeProp !== void 0 ? typeProp : 'confirm';
  const icon = (() => {
    /**
     * 现实效果: 用户传入icon: null时 icon不显示 (当用户不穿入icon属性时 icon为undefined)
     * 使用isNonNullable函数 -> iconProp为 null 与 undefined 时均根据typeProp选择相应icon
     * 此处对用户不传iconProp 和 传入iconProp: null分开处理
     */
    if (iconProp !== undefined) {
      return iconProp;
    }
    switch (typeProp) {
      case 'error':
        return /*#__PURE__*/React.createElement(ErrorFilled, null);
      case 'success':
        return /*#__PURE__*/React.createElement(SucceedFilled, null);
      case 'info':
        return /*#__PURE__*/React.createElement(InfoFilled, null);
      case 'confirm':
      case 'warning':
        return /*#__PURE__*/React.createElement(WarningFilled, null);
      default:
        return null;
    }
  })();
  const cssScope = useCssScope('confirm');
  const locale = useLocale();
  const cancelRef = React.useRef(null);
  React.useImperativeHandle(cancelButtonRef, () => cancelRef.current);
  const okRef = React.useRef(null);
  React.useImperativeHandle(okButtonRef, () => okRef.current);
  const [okLoading, setOkLoading] = useState(Boolean(okButtonProps === null || okButtonProps === void 0 ? void 0 : okButtonProps.loading));
  const [cancelLoading, setCancelLoading] = useState(Boolean(cancelButtonProps === null || cancelButtonProps === void 0 ? void 0 : cancelButtonProps.loading));
  const triggerClose = () => onClose && onClose();

  // 自动设置焦点，回车后会触发相关事件
  useEffectOnce(() => {
    let timerId;
    if (autoFocusButton) {
      timerId = window.setTimeout(() => {
        var _cancelRef$current, _okRef$current;
        autoFocusButton === 'cancel' ? (_cancelRef$current = cancelRef.current) === null || _cancelRef$current === void 0 ? void 0 : _cancelRef$current.focus() : (_okRef$current = okRef.current) === null || _okRef$current === void 0 ? void 0 : _okRef$current.focus();
      });
    }
    return () => {
      if (timerId) {
        window.clearTimeout(timerId);
      }
    };
  });
  useEffect(() => {
    setOkLoading(Boolean(okButtonProps === null || okButtonProps === void 0 ? void 0 : okButtonProps.loading));
    setCancelLoading(Boolean(cancelButtonProps === null || cancelButtonProps === void 0 ? void 0 : cancelButtonProps.loading));
  }, [okButtonProps, cancelButtonProps]);
  const handleOnConfirmClick = event => {
    if (okButtonProps && okButtonProps.onClick) {
      okButtonProps.onClick(event);
    }
    if (!onOk) {
      triggerClose();
    } else {
      if (onOk.length) {
        onOk(triggerClose);
      } else {
        const result = onOk();
        if (result && result.then) {
          setOkLoading(true);
          result.then(() => {
            setOkLoading(false);
            triggerClose();
          }, e => {
            console.error(e);
            setOkLoading(false);
          });
        } else {
          triggerClose();
        }
      }
    }
  };
  const handleOnCancelClick = event => {
    if (cancelButtonProps && cancelButtonProps.onClick) {
      cancelButtonProps.onClick(event);
    }
    if (!onCancel) {
      triggerClose();
    } else {
      if (onCancel.length) {
        onCancel(triggerClose);
      } else {
        const result = onCancel();
        if (result && result.then) {
          setCancelLoading(true);
          result.then(() => {
            setCancelLoading(false);
            triggerClose();
          }, e => {
            console.error(e);
            setCancelLoading(false);
          });
        } else {
          triggerClose();
        }
      }
    }
  };
  const headerNode = /*#__PURE__*/React.createElement("div", {
    className: cssScope('__header')
  }, /*#__PURE__*/React.createElement(FirstLineAlign, {
    className: cssScope('__title'),
    sx: [cssTypographyTitle4],
    beforeProps: isNonNullable(icon) ? {
      className: cssScope('__icon'),
      sx: [sxDialogIconWrapper],
      children: /*#__PURE__*/React.createElement("div", {
        className: stylex(sxDialogIcon)
      }, icon)
    } : undefined
  }, /*#__PURE__*/React.createElement("div", {
    className: cssScope('__titleContent')
  }, title)), isNonNullable(content) ? /*#__PURE__*/React.createElement("div", {
    style: bodyStyle,
    className: clsx(cssScope('__body'), stylex(cssTypographyBody0))
  }, content) : null);
  return /*#__PURE__*/React.createElement("div", {
    className: clsx(className, cssScope('__content'), {
      [cssScope('-hasIcon')]: icon,
      [cssScope(`-${type}`)]: true
    }, stylex(sxModalContent.root, cssTypographyBody0, ...sx)),
    style: style
  }, headerNode, (() => /*#__PURE__*/React.createElement("div", {
    className: cssScope('__footer')
  }, extraAction ? /*#__PURE__*/React.createElement("div", {
    className: cssScope('__footer__extraAction')
  }, extraAction) : null, /*#__PURE__*/React.createElement("div", {
    className: cssScope('__footer__btns')
  }, /*#__PURE__*/React.createElement(Button, _extends({
    ref: okRef,
    type: "primary"
  }, okButtonProps, {
    onClick: handleOnConfirmClick,
    loading: okLoading
  }), okText || locale.components.modal.okText), enableCancelButton && /*#__PURE__*/React.createElement(Button, _extends({
    ref: cancelRef
  }, cancelButtonProps, {
    loading: cancelLoading,
    onClick: handleOnCancelClick
  }), cancelText || locale.components.modal.cancelText))))());
};

// eslint-disable-next-line max-lines-per-function
export const Confirm = props => {
  const {
    icon,
    title,
    content,
    keyboardClosable = true,
    mask = true,
    centered = true,
    destroyOnClose,
    visible,
    cancelText,
    okText,
    cancelButtonProps,
    okButtonProps,
    enableCancelButton = true,
    maskStyle,
    bodyStyle,
    autoFocusButton = 'ok',
    onCancel,
    onOk,
    onClose,
    // 内部属性
    onAfterOpen,
    onAfterClose,
    className,
    style,
    wrapClassName,
    type = 'confirm',
    extraAction,
    maskClosable = false,
    ...restProps
  } = props;
  const cssScope = useCssScope('confirm');
  const triggerClose = () => onClose && onClose();
  const initialFocusRef = React.useRef(null);
  return /*#__PURE__*/React.createElement(Dialog, _extends({
    visible: visible,
    centered: centered,
    onClose: () => {
      triggerClose();
      isFunction(onCancel) && onCancel();
    },
    overlay: mask,
    onAfterOpen: onAfterOpen,
    onAfterClose: () => {
      isFunction(onAfterClose) && onAfterClose();
    },
    destroyOnClose: destroyOnClose,
    overlayStyle: maskStyle,
    wrapperClassName: wrapClassName,
    contentStyle: style,
    overlayClosable: maskClosable,
    keyboardClosable: keyboardClosable,
    contentClassName: clsx(cssScope(), className),
    size: "small",
    initialFocusRef: autoFocusButton ? initialFocusRef : undefined
  }, restProps), /*#__PURE__*/React.createElement(ConfirmContent, {
    icon,
    title,
    content,
    cancelText,
    okText,
    cancelButtonProps,
    okButtonProps,
    enableCancelButton,
    autoFocusButton,
    bodyStyle,
    onCancel,
    onOk,
    onClose,
    type,
    extraAction,
    sx: [centered && {
      "props": {
        "B6of3ja": uwejdxd,
        "t21cq0": u1purjcd,
        "jrapky": uuhcnxi,
        "Frg6f3": u2o5q4q
      }
    }],
    okButtonRef: autoFocusButton === 'ok' ? initialFocusRef : undefined,
    cancelButtonRef: autoFocusButton === 'cancel' ? initialFocusRef : undefined
  }));
};
export function withWarn(props) {
  return {
    enableCancelButton: false,
    ...props,
    type: 'warning'
  };
}
export function withInfo(props) {
  return {
    enableCancelButton: false,
    ...props,
    type: 'info'
  };
}
export function withSuccess(props) {
  return {
    enableCancelButton: false,
    ...props,
    type: 'success'
  };
}
export function withError(props) {
  return {
    enableCancelButton: false,
    ...props,
    type: 'error'
  };
}
export function withConfirm(props) {
  return {
    enableCancelButton: true,
    ...props,
    type: 'confirm'
  };
}
