// @universe-design/react/es/dialog/index.js
import React from 'react';
import ReactDOM from 'react-dom';
import { Confirm, withConfirm, withError, withInfo, withSuccess, withWarn } from './Confirm';
import { Modal as ModalComponents, ModalContent } from './Modal';
import { isFunction } from '../util';
import { CrossRootConfigProvider } from '../config-provider';
const destroyMap = new Map();
export const Dialog = ModalComponents;
function confirm(_ref) {
  let {
    onAfterClose,
    visible,
    ...config
  } = _ref;
  const uuid = `ud-confirm-instance-${new Date().getTime()}`;
  const container = document.createElement('div');
  document.body.appendChild(container);
  let currentConfig;
  function destroy() {
    const unmountResult = ReactDOM.unmountComponentAtNode(container);
    if (unmountResult && container.parentNode) {
      container.parentNode.removeChild(container);
    }
    // config.onCancel && config.onCancel();
    destroyMap.delete(uuid);
  }
  function render(renderConfig) {
    ReactDOM.render(/*#__PURE__*/React.createElement(CrossRootConfigProvider, null, /*#__PURE__*/React.createElement(Confirm, renderConfig)), container);
  }
  function close() {
    currentConfig = {
      ...currentConfig,
      visible: false
    };
    render(currentConfig);
  }
  function update(newConfig) {
    if (isFunction(newConfig)) {
      currentConfig = newConfig(currentConfig);
    } else {
      currentConfig = {
        ...currentConfig,
        ...newConfig
      };
    }
    render(currentConfig);
  }
  currentConfig = {
    ...config,
    onAfterClose: () => {
      isFunction(onAfterClose) && onAfterClose();
      destroy();
    },
    visible: true,
    onClose: close
  };
  render(currentConfig);
  destroyMap.set(uuid, destroy);
  return {
    destroy: close,
    update
  };
}
Dialog.destroyAll = () => {
  destroyMap.forEach(fn => fn());
  destroyMap.clear();
};
Dialog.Content = ModalContent;
Dialog.confirm = config => confirm(withConfirm(config));
Dialog.info = config => confirm(withInfo(config));
Dialog.success = config => confirm(withSuccess(config));
Dialog.error = config => confirm(withError(config));
Dialog.warning = config => confirm(withWarn(config));
export { ConfirmContent, Confirm } from './Confirm';
export { useDialog } from './hooks/useDialog';
export { DialogCancelType, DialogLayoutMode } from './dialog';
