// @universe-design/react/es/dialog/Modal.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.uwegl26{box-shadow:@shadow-s5-down;}
import um9xw6y from "./../.stylex/um9xw6y";
//.ugr6219{margin-top:auto;}
import uwejdxd from "./../.stylex/uwejdxd";
//.u11qmguv{margin-right:0;}
import u1purjcd from "./../.stylex/u1purjcd";
//.u10jk5vf{margin-bottom:auto;}
import uuhcnxi from "./../.stylex/uuhcnxi";
//.u1tyq0we{margin-left:0;}
import u2o5q4q from "./../.stylex/u2o5q4q";
//.u1a3p1vp{overflow:hidden;}
import ux645fm from "./../.stylex/ux645fm";
/* eslint-disable @typescript-eslint/no-use-before-define */

import React, { useCallback, useEffect, useState, forwardRef } from 'react';
import clsx from 'clsx';
import stylex from '@universe-design/stylex';
import { cssTypographyTitle4, cssTypographyBody0 } from '../stylex';
import { Dialog, DialogCancelType, useDialogLayoutMode } from './dialog';
import { Button } from '../button';
import CloseOutlined from '@universe-design/icons-react/CloseOutlined';
import { useCssScope, useLocale } from '../config-provider/hooks';
import { useSize } from '../hooks';
import { useResizeObserver, useCloneElementRef } from '@universe-design/react-utilities';
import { isFunction, isNonNullable } from '../util';
import { getThunkElement } from '../util/getThunkElement';
import { FirstLineAlign } from '../shared/components/FirstLineAlign';
import { sxModalContent } from './sx';
import './style';
function useCheckVerticalOverflow(_ref) {
  let {
    target
  } = _ref;
  const [overflowTop, setOverflowTop] = useState(false);
  const [overflowBottom, setOverflowBottom] = useState(false);
  const checkVerticalOverflow = useCallback(() => {
    const targetEl = getThunkElement(target);
    if (!targetEl) {
      return;
    }
    const {
      scrollTop,
      scrollHeight,
      clientHeight
    } = targetEl;
    const targetInset = {
      top: 0,
      bottom: clientHeight
    };
    const childInset = {
      top: -scrollTop,
      bottom: -scrollTop + scrollHeight
    };
    if (targetInset.top > childInset.top) {
      setOverflowTop(true);
    } else {
      setOverflowTop(false);
    }
    if (targetInset.bottom < childInset.bottom) {
      setOverflowBottom(true);
    } else {
      setOverflowBottom(false);
    }
  }, [target]);
  return {
    overflow: {
      top: overflowTop,
      bottom: overflowBottom
    },
    checkVerticalOverflow
  };
}

/**
 * @public
 * @docgen
 * <Dialog.Content /> 的属性。
 */

/**
 * @public
 * @docgen
 * <Dialog /> 的属性。
 */

export const ModalContent = /*#__PURE__*/forwardRef((props, ref) => {
  const cssScope = useCssScope('modal');
  const {
    header,
    title,
    description,
    closable = true,
    enableCancelButton = true,
    children,
    cancelText,
    okText,
    cancelButtonProps,
    okButtonProps,
    bodyStyle = {},
    confirmLoading,
    footer,
    onCancel,
    onOk,
    extraAction,
    style,
    className,
    sx = [],
    bodyHeightResponsiveMode = 'scroll',
    maxHeight,
    ...restProps
  } = props;
  const locale = useLocale();
  const layoutMode = useDialogLayoutMode();
  const closeIconNode = closable ? /*#__PURE__*/React.createElement(Button, {
    className: cssScope('__close'),
    onClick: e => {
      isFunction(onCancel) && onCancel(e, {
        cancelType: DialogCancelType.CloseIcon
      });
    },
    icon: /*#__PURE__*/React.createElement(CloseOutlined, null),
    type: "icon"
  }) : null;
  const bodyScrollable = bodyHeightResponsiveMode === 'scroll';
  const bodyStretch = bodyHeightResponsiveMode === 'stretch';
  const [headerSizeRef, headerSize] = useSize({
    resizeObserverDisabled: !bodyScrollable && !bodyStretch
  });
  const [footerSizeRef, footerSize] = useSize({
    resizeObserverDisabled: !bodyScrollable && !bodyStretch
  });
  const contentCls = clsx(className, cssScope('__content'), closeIconNode && cssScope('-closable'), stylex(sxModalContent.root, bodyScrollable && sxModalContent.scrollable, cssTypographyBody0, bodyHeightResponsiveMode === 'stretch' && sxModalContent.stretch, ...sx));
  const handleCancel = e => {
    if (cancelButtonProps && cancelButtonProps.onClick) {
      cancelButtonProps.onClick(e);
    }
    isFunction(onCancel) && onCancel(e, {
      cancelType: DialogCancelType.CancelButton
    });
  };
  const handleOk = e => {
    if (okButtonProps && okButtonProps.onClick) {
      okButtonProps.onClick(e);
    }
    isFunction(onOk) && onOk(e);
  };

  /**
   * Title ReactNode
   */
  let headerNode = null;
  if (header !== null) {
    headerNode = /*#__PURE__*/React.createElement("div", {
      ref: headerSizeRef,
      className: cssScope('__header')
    }, isNonNullable(header) ? header : /*#__PURE__*/React.createElement(FirstLineAlign, {
      className: cssScope('__title'),
      sx: [cssTypographyTitle4]
    }, /*#__PURE__*/React.createElement("div", {
      className: cssScope('__titleContent')
    }, title)), isNonNullable(description) ? /*#__PURE__*/React.createElement("div", {
      className: clsx(cssScope('__des'), stylex(cssTypographyBody0))
    }, description) : null);
  }

  /**
   * Footer ReactNode
   */
  const footerNode = (() => {
    const footerContentNode = /*#__PURE__*/React.createElement(React.Fragment, null, extraAction ? /*#__PURE__*/React.createElement("div", {
      className: cssScope('__footer__extraAction')
    }, extraAction) : null, /*#__PURE__*/React.createElement("div", {
      className: cssScope('__footer__btns')
    }, /*#__PURE__*/React.createElement(Button, _extends({
      type: "primary"
    }, okButtonProps, {
      loading: confirmLoading || (okButtonProps === null || okButtonProps === void 0 ? void 0 : okButtonProps.loading),
      onClick: handleOk
    }), okText || locale.components.modal.okText), enableCancelButton && /*#__PURE__*/React.createElement(Button, _extends({}, cancelButtonProps, {
      onClick: handleCancel
    }), cancelText || locale.components.modal.cancelText)));
    if (footer) {
      return /*#__PURE__*/React.createElement("div", {
        ref: footerSizeRef,
        className: cssScope('__footer')
      }, footer);
    }
    if (footer === null) {
      return null;
    }
    return /*#__PURE__*/React.createElement("div", {
      ref: footerSizeRef,
      className: cssScope('__footer')
    }, footerContentNode);
  })();

  /**
   * Body ReactNode
   */

  // body 内容超出增加分割线
  const bodyRef = useResizeObserver(() => {
    checkVerticalOverflow();
  });
  const {
    overflow: bodyOverflow,
    checkVerticalOverflow
  } = useCheckVerticalOverflow({
    target: bodyRef
  });
  const childrenRef = useResizeObserver(() => {
    checkVerticalOverflow();
  });
  useEffect(() => {
    const body = bodyRef.current;
    if (!body) {
      return;
    }
    body.addEventListener('scroll', checkVerticalOverflow);
    return () => {
      body.removeEventListener('scroll', checkVerticalOverflow);
    };
  }, [bodyRef, checkVerticalOverflow]);
  const childrenMaybeWithRef = useCloneElementRef({
    children,
    ref: childrenRef
  });
  const {
    height: bodyHeight
  } = bodyStyle;
  const bodyMinHeight = (() => {
    if (isNonNullable(bodyStyle.minHeight)) {
      return bodyStyle.minHeight;
    }
    if (layoutMode === 'legacy') {
      return 0;
    }
    // 62 是 body 的最小高度，一个字段 32 + 8 + 22
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    return 32 + 8 + 22;
  })();
  const bodyNode = isNonNullable(children) ? /*#__PURE__*/React.createElement('div', {
    style: {
      ...bodyStyle,
      minHeight: bodyMinHeight,
      overflowY: bodyScrollable ? 'auto' : undefined
    },
    className: clsx(cssScope('__body'), headerNode && bodyOverflow.top && cssScope('__body--top-overflow'), footerNode && bodyOverflow.bottom && cssScope('__body--bottom-overflow'), stylex(bodyStretch && {
      "props": {
        "Bpg54ce": ux645fm
      }
    })),
    ref: bodyRef
  }, childrenMaybeWithRef) : null;
  const contentStyle = (() => {
    // 24 目前在 body 底部的 margin

    const sharedStyle = {
      ...style,
      maxHeight
    };
    if (bodyScrollable || bodyStretch) {
      const bodyMinHeightStr = (() => {
        if (!bodyNode) {
          return;
        }
        if (isNonNullable(bodyHeight)) {
          return typeof bodyHeight === 'number' ? `${bodyHeight}px` : bodyHeight;
        }
        return typeof bodyMinHeight === 'number' ? `${bodyMinHeight}px` : bodyMinHeight;
      })();
      const minHeight = `calc(${[`${headerSize.height + footerSize.height + (bodyNode ? 24 : 0)}px`, bodyMinHeightStr].filter(Boolean).join(' + ')})`;
      return {
        ...sharedStyle,
        minHeight
      };
    }
    return sharedStyle;
  })();
  return /*#__PURE__*/React.createElement("div", _extends({
    className: contentCls,
    style: contentStyle,
    ref: ref
  }, restProps), headerNode, closeIconNode, bodyNode, footerNode);
});

// eslint-disable-next-line max-lines-per-function
export const Modal = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    title,
    zIndex,
    description,
    closable = true,
    enableCancelButton = true,
    maskClosable = true,
    keyboardClosable = true,
    centered = true,
    width,
    visible,
    children,
    cancelText,
    okText,
    cancelButtonProps,
    okButtonProps,
    mask = true,
    style,
    maskStyle,
    bodyStyle,
    enableAnimation,
    transitionName,
    maskTransitionName,
    confirmLoading,
    // SSR compatiable
    getContainer,
    footer,
    onCancel,
    onOk,
    onAfterOpen,
    onAfterClose,
    className,
    wrapClassName,
    extraAction,
    size = 'medium',
    header,
    safeMargin,
    onClose,
    bodyHeightResponsiveMode,
    maxHeight,
    ...restProps
  } = props;
  const cssScope = useCssScope('modal');
  const layoutMode = useDialogLayoutMode();
  return /*#__PURE__*/React.createElement(Dialog, _extends({
    ref: ref,
    width: width,
    zIndex: zIndex,
    visible: visible,
    centered: centered,
    overlay: mask,
    overlayStyle: maskStyle,
    wrapperClassName: wrapClassName,
    overlayClosable: maskClosable,
    contentStyle: style,
    contentClassName: clsx(cssScope(), className),
    keyboardClosable: keyboardClosable,
    enableAnimation: enableAnimation,
    overlayTransitionClassName: transitionName,
    contentTransitionClassName: maskTransitionName,
    getContainer: getContainer,
    onClose: (e, payload) => {
      isFunction(onCancel) && onCancel(e, payload);
      if (e !== null && e !== void 0 && e.isDefaultPrevented()) {
        return;
      }
      // 这个逻辑只给 imperative dialog 使用
      isFunction(onClose) && onClose();
    },
    onAfterOpen: onAfterOpen,
    onAfterClose: onAfterClose,
    size: size,
    safeMargin: safeMargin
  }, restProps), /*#__PURE__*/React.createElement(ModalContent, {
    header,
    title,
    description,
    closable,
    enableCancelButton,
    children,
    cancelText,
    okText,
    cancelButtonProps,
    okButtonProps,
    bodyStyle,
    confirmLoading,
    footer,
    onCancel,
    onOk,
    extraAction,
    sx: [layoutMode === 'legacy' ? [] : [centered && {
      "props": {
        "B6of3ja": uwejdxd,
        "t21cq0": u1purjcd,
        "jrapky": uuhcnxi,
        "Frg6f3": u2o5q4q
      }
    }, !mask && {
      "props": {
        "E5pizo": um9xw6y
      }
    }]],
    bodyHeightResponsiveMode,
    maxHeight
  }));
});
