// @universe-design/react/es/dialog/sx.js
//.u1l02sjl{height:100%;}
import udg2fzc from "./../.stylex/udg2fzc";
//.u1immsc2{max-height:100%;}
import u1j86i34 from "./../.stylex/u1j86i34";
//.u10pi13n{position:relative;}
import u5jxv71 from "./../.stylex/u5jxv71";
//.u1vx9l62{flex-direction:column;}
import uohwdt from "./../.stylex/uohwdt";
//.uly5x3f{width:100%;}
import u13d3rh8 from "./../.stylex/u13d3rh8";
//.u1rt2boy{font-size:24px;}
import udavndu from "./../.stylex/udavndu";
//.uez10in{line-height:0;}
import u5p8ukz from "./../.stylex/u5p8ukz";
//.u7nlbp4{align-self:center;}
import u1mtnbey from "./../.stylex/u1mtnbey";
//.u17vyym1{margin-right:16px;}
import u1hcffm1 from "./../.stylex/u1hcffm1";
import stylex from '@universe-design/stylex';
import { cssFlexBase, sxBorderRadius } from '../stylex';

/**
 * tokens
 */
const DialogIconSize = 24;
const DialogIconTitleSpacing = 16;
export const sxDialogIconWrapper = {
  "props": {
    "t21cq0": u1hcffm1
  }
};
export const sxDialogIcon = {
  "props": {
    "Be2twd7": udavndu,
    "Bg96gwp": u5p8ukz,
    "qb2dma": u1mtnbey
  }
};
export const sxModalContent = {
  root: stylex.compose(cssFlexBase, sxBorderRadius.md, {
    "props": {
      "qhf8xq": u5jxv71,
      "Beiy3e4": uohwdt,
      "a9b677": u13d3rh8
    }
  }),
  scrollable: {
    "props": {
      "Bxyxcbc": u1j86i34
    }
  },
  stretch: {
    "props": {
      "Bqenvij": udg2fzc
    }
  }
};
