// @universe-design/react/es/breadcrumb/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React, { cloneElement, useState } from 'react';
import clsx from 'clsx';
import MoreOutlined from '@universe-design/icons-react/MoreOutlined';
import RightBoldOutlined from '@universe-design/icons-react/RightBoldOutlined';
import { useCssScope } from '../config-provider/hooks';
import { BreadcrumbItem } from './BreadcrumbItem';
import { Popover } from '../popover';
import { Overflow } from '../overflow';
import { toFlatArray } from '../util';
import './style';
export const Breadcrumb = Object.assign(/*#__PURE__*/React.forwardRef((props, ref) => {
  const {
    className,
    children,
    maxCount = Infinity,
    separator = /*#__PURE__*/React.createElement(RightBoldOutlined, null),
    ...restProps
  } = props;
  const cssScope = useCssScope('breadcrumb');
  const childNodes = toFlatArray(children);
  const lastChildNode = childNodes.pop();
  const separatorNode = /*#__PURE__*/React.createElement("span", {
    className: cssScope('-separator')
  }, separator);
  const [visible, setVisible] = useState(false);
  return /*#__PURE__*/React.createElement("div", _extends({
    className: clsx(cssScope(), className),
    ref: ref
  }, restProps), /*#__PURE__*/React.createElement(Overflow, {
    restPosition: 1,
    maxCount: typeof maxCount === 'number' ? maxCount - 1 : 'responsive',
    keys: Array(childNodes.length).fill(0).map((_, i) => i),
    itemRenderer: key => /*#__PURE__*/React.createElement(React.Fragment, null, childNodes[key], separatorNode),
    restRenderer: omittedKeys => /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Popover, {
      mode: "pure",
      visible: visible,
      content: /*#__PURE__*/React.createElement("div", {
        onClick: () => setVisible(false),
        className: cssScope('__popover__content')
      }, omittedKeys.map((o, i) => /*#__PURE__*/React.createElement(React.Fragment, {
        key: o
      }, childNodes[o], i !== omittedKeys.length - 1 && separatorNode))),
      placement: "bottomLeft",
      hideArrow: true,
      onVisibleChange: setVisible
    }, /*#__PURE__*/React.createElement(BreadcrumbItem, {
      className: cssScope('-ellipses')
    }, /*#__PURE__*/React.createElement(MoreOutlined, null))), separatorNode),
    suffixRenderer: () => lastChildNode ? /*#__PURE__*/cloneElement(lastChildNode, {
      interactive: false
    }) : null
  }));
}), {
  Item: BreadcrumbItem
});
if (process.env.NODE_ENV !== "production") {
  Breadcrumb.displayName = 'Breadcrumb';
}
