// @universe-design/react/lib/config-provider/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  CrossRootConfigProvider: true,
  ConfigProvider: true,
  Consumer: true,
  COMPONENT_VERSION_ID: true,
  ConfigContext: true,
  useComponentSize: true,
  useInputGroupSize: true,
  useCssScope: true,
  useLocale: true,
  useClassPrefix: true,
  useLocaleReceiver: true,
  useDirection: true,
  useGetTargetContainer: true,
  useTheme: true
};
exports.COMPONENT_VERSION_ID = void 0;
Object.defineProperty(exports, "ConfigContext", {
  enumerable: true,
  get: function () {
    return _ConfigContext.ConfigContext;
  }
});
exports.CrossRootConfigProvider = exports.Consumer = exports.ConfigProvider = void 0;
Object.defineProperty(exports, "useClassPrefix", {
  enumerable: true,
  get: function () {
    return _hooks.useClassPrefix;
  }
});
Object.defineProperty(exports, "useComponentSize", {
  enumerable: true,
  get: function () {
    return _hooks.useComponentSize;
  }
});
Object.defineProperty(exports, "useCssScope", {
  enumerable: true,
  get: function () {
    return _hooks.useCssScope;
  }
});
Object.defineProperty(exports, "useDirection", {
  enumerable: true,
  get: function () {
    return _hooks.useDirection;
  }
});
Object.defineProperty(exports, "useGetTargetContainer", {
  enumerable: true,
  get: function () {
    return _hooks.useGetTargetContainer;
  }
});
Object.defineProperty(exports, "useInputGroupSize", {
  enumerable: true,
  get: function () {
    return _hooks.useInputGroupSize;
  }
});
Object.defineProperty(exports, "useLocale", {
  enumerable: true,
  get: function () {
    return _hooks.useLocale;
  }
});
Object.defineProperty(exports, "useLocaleReceiver", {
  enumerable: true,
  get: function () {
    return _hooks.useLocaleReceiver;
  }
});
Object.defineProperty(exports, "useTheme", {
  enumerable: true,
  get: function () {
    return _hooks.useTheme;
  }
});
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _ThemeProvider = require("@universe-design/icons-react/ThemeProvider");
var _reactProvider = require("@universe-design/react-provider");
var _useContextSelector = require("use-context-selector");
var _shim = require("use-sync-external-store/shim");
var _mergeWith = _interopRequireDefault(require("lodash/mergeWith"));
var _uniq = _interopRequireDefault(require("lodash/uniq"));
var _miniStore = require("./miniStore");
var _reactUtilities = require("@universe-design/react-utilities");
var _useRenderEffect = require("../hooks/useRenderEffect");
var _ConfigContext = require("./ConfigContext");
var _constants = require("./constants");
require("./style");
var _desktop = require("./desktop");
Object.keys(_desktop).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _desktop[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _desktop[key];
    }
  });
});
var _hooks = require("./hooks");
// ConfigProvider，几种情况
// 1. CrossRootConfigProvider 或者 ConfigProvider 如果有父级，则直接用父级和 props merge 得到当前层的结果
// 2. CrossRootConfigProvider 如果没有父级，则使用 module scope 的 config 作为父级
// 3. ConfigProvider 如果没有父级，则需要将当前层级的 config 更新到 module scope config

/**
 * `ConfigProvider` 可嵌套，props 可传递，用于解决 light 中配置 dark 的情况。
 * `useLocaleFontFamily`和对`miniStore`操作全局唯一。
 */

const ConfigChildren = props => {
  const {
    parentContext = _constants.defaultConfig,
    isOutermost,
    // props
    locale = parentContext.locale,
    prefixCls = parentContext.rootPrefixCls,
    direction = parentContext.direction,
    getTargetContainer = parentContext.getTargetContainer,
    toast = parentContext.toast,
    theme = parentContext.theme,
    textOverflowPopperSidecar = parentContext.textOverflowPopperSidecar,
    useLocaleFontFamily = false,
    componentSize = parentContext.componentSize,
    children,
    PopperSidecar = parentContext.PopperSidecar,
    timezone = parentContext.timezone
  } = props;

  // 处理 partial locale
  const finalLocale = React.useMemo(() =>
  // 递归合并 defaultLocale 和 locale，但遇到数组时只进行浅覆盖
  (0, _mergeWith.default)({}, parentContext.locale, locale, (objValue, srcValue) => {
    if (Array.isArray(objValue)) {
      return srcValue;
    }
  }), [locale, parentContext.locale]);
  const getPopupContainer = (0, _reactProvider.usePopupPortalContainer)();
  const config = React.useMemo(() => ({
    rootPrefixCls: prefixCls,
    locale: finalLocale,
    direction,
    getPopupContainer,
    getTargetContainer,
    toast,
    theme,
    componentSize,
    textOverflowPopperSidecar,
    PopperSidecar,
    timezone
  }), [prefixCls, finalLocale, direction, getPopupContainer, getTargetContainer, toast, theme, componentSize, textOverflowPopperSidecar, PopperSidecar, timezone]);
  (0, _useRenderEffect.useRenderEffect)(React.useCallback(() => {
    if (!isOutermost || !(0, _reactUtilities.isClient)()) {
      return;
    }
    if (!(useLocaleFontFamily && locale.locale)) {
      return;
    }
    const lang = locale.locale.split('-')[0];
    if (!lang) {
      return;
    }
    const localeFontClassName = `${prefixCls}__lang-${lang}`;
    const getClassNames = () => document.body.className.split(/\s+/);
    document.body.className = (0, _uniq.default)([...getClassNames(), localeFontClassName]).join(' ');
    return () => {
      document.body.className = getClassNames().filter(o => o !== localeFontClassName).join(' ');
    };
  }, [isOutermost, locale.locale, prefixCls, useLocaleFontFamily]));
  React.useEffect(() => {
    if (!isOutermost) {
      return;
    }
    _miniStore.configStore.set(config);
  }, [isOutermost, config]);
  return /*#__PURE__*/React.createElement(_ThemeProvider.ThemeProvider, {
    mode: config.theme
  }, /*#__PURE__*/React.createElement(_ConfigContext.ConfigContext.Provider, {
    value: config
  }, children));
};

/**
 * 内部使用
 * @internal
 */
const CrossRootConfigProvider = props => {
  const {
    children
  } = props;
  const globalConfig = (0, _shim.useSyncExternalStore)(_miniStore.configStore.subscribe, _miniStore.configStore.get, _miniStore.configStore.get);
  const child = /*#__PURE__*/React.createElement(ConfigChildren, {
    parentContext: globalConfig,
    isOutermost: false
  }, children);
  if (globalConfig.getPopupContainer) {
    return /*#__PURE__*/React.createElement(_reactProvider.PopupPortalContainerProvider, {
      value: globalConfig.getPopupContainer
    }, child);
  }
  return child;
};

/**
 *
 * <CrossRootConfigProvider>
 *   <ConfigChildren />
 * </CrossRootConfigProvider>
 *
 * <ConfigProvider>
 *   <ConfigChildren />
 * </ConfigProvider>
 */
exports.CrossRootConfigProvider = CrossRootConfigProvider;
const ConfigProvider = props => {
  const parentContext = (0, _useContextSelector.useContext)(_ConfigContext.ConfigContext);
  const {
    getPopupContainer,
    ...restProp
  } = props;
  const child = /*#__PURE__*/React.createElement(ConfigChildren, (0, _extends2.default)({
    parentContext: parentContext,
    isOutermost: parentContext === _constants.defaultConfig
  }, restProp));
  if (getPopupContainer) {
    return /*#__PURE__*/React.createElement(_reactProvider.PopupPortalContainerProvider, {
      value: getPopupContainer
    }, child);
  }
  return child;
};
exports.ConfigProvider = ConfigProvider;
const Consumer = props => {
  const config = (0, _useContextSelector.useContext)(_ConfigContext.ConfigContext);
  return props.children({
    ...config
  });
};
exports.Consumer = Consumer;
const COMPONENT_VERSION_ID = exports.COMPONENT_VERSION_ID = '__VERSION__';
