// @universe-design/react/lib/shared/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  Highlighter: true,
  HighlightConsumer: true,
  HighlightProvider: true,
  FocusLock: true,
  ScrollArea: true,
  useThemeMode: true,
  useDerivedThemeModeState: true,
  getSystemThemeMode: true,
  importLocale: true
};
Object.defineProperty(exports, "FocusLock", {
  enumerable: true,
  get: function () {
    return _components.FocusLock;
  }
});
Object.defineProperty(exports, "HighlightConsumer", {
  enumerable: true,
  get: function () {
    return _components.HighlightConsumer;
  }
});
Object.defineProperty(exports, "HighlightProvider", {
  enumerable: true,
  get: function () {
    return _components.HighlightProvider;
  }
});
Object.defineProperty(exports, "Highlighter", {
  enumerable: true,
  get: function () {
    return _components.Highlighter;
  }
});
Object.defineProperty(exports, "ScrollArea", {
  enumerable: true,
  get: function () {
    return _components.ScrollArea;
  }
});
Object.defineProperty(exports, "getSystemThemeMode", {
  enumerable: true,
  get: function () {
    return _useThemeMode.getSystemThemeMode;
  }
});
Object.defineProperty(exports, "importLocale", {
  enumerable: true,
  get: function () {
    return _importLocale.importLocale;
  }
});
Object.defineProperty(exports, "useDerivedThemeModeState", {
  enumerable: true,
  get: function () {
    return _useThemeMode.useDerivedThemeModeState;
  }
});
Object.defineProperty(exports, "useThemeMode", {
  enumerable: true,
  get: function () {
    return _useThemeMode.useThemeMode;
  }
});
var _components = require("./components");
var _popper = require("./components/popper");
Object.keys(_popper).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _popper[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _popper[key];
    }
  });
});
var _FloatingUIPopperSidecar = require("./components/popper-sidecar/FloatingUIPopperSidecar");
Object.keys(_FloatingUIPopperSidecar).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _FloatingUIPopperSidecar[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _FloatingUIPopperSidecar[key];
    }
  });
});
var _RcTriggerPopperSidecar = require("./components/popper-sidecar/RcTriggerPopperSidecar");
Object.keys(_RcTriggerPopperSidecar).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _RcTriggerPopperSidecar[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _RcTriggerPopperSidecar[key];
    }
  });
});
var _locales = require("./locales");
Object.keys(_locales).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _locales[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _locales[key];
    }
  });
});
var _useThemeMode = require("./hooks/useThemeMode");
var _importLocale = require("./importLocale");
