// @universe-design/react/lib/shared/locales.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  defaultLocale: true,
  deDE: true,
  enUS: true,
  esES: true,
  frFR: true,
  idID: true,
  itIT: true,
  jaJP: true,
  koKR: true,
  ptBR: true,
  ruRU: true,
  thTH: true,
  viVN: true,
  zhCN: true,
  zhHK: true,
  zhTW: true
};
Object.defineProperty(exports, "deDE", {
  enumerable: true,
  get: function () {
    return _deDE.default;
  }
});
exports.defaultLocale = void 0;
Object.defineProperty(exports, "enUS", {
  enumerable: true,
  get: function () {
    return _enUS.default;
  }
});
Object.defineProperty(exports, "esES", {
  enumerable: true,
  get: function () {
    return _esES.default;
  }
});
Object.defineProperty(exports, "frFR", {
  enumerable: true,
  get: function () {
    return _frFR.default;
  }
});
Object.defineProperty(exports, "idID", {
  enumerable: true,
  get: function () {
    return _idID.default;
  }
});
Object.defineProperty(exports, "itIT", {
  enumerable: true,
  get: function () {
    return _itIT.default;
  }
});
Object.defineProperty(exports, "jaJP", {
  enumerable: true,
  get: function () {
    return _jaJP.default;
  }
});
Object.defineProperty(exports, "koKR", {
  enumerable: true,
  get: function () {
    return _koKR.default;
  }
});
Object.defineProperty(exports, "ptBR", {
  enumerable: true,
  get: function () {
    return _ptBR.default;
  }
});
Object.defineProperty(exports, "ruRU", {
  enumerable: true,
  get: function () {
    return _ruRU.default;
  }
});
Object.defineProperty(exports, "thTH", {
  enumerable: true,
  get: function () {
    return _thTH.default;
  }
});
Object.defineProperty(exports, "viVN", {
  enumerable: true,
  get: function () {
    return _viVN.default;
  }
});
Object.defineProperty(exports, "zhCN", {
  enumerable: true,
  get: function () {
    return _zhCN.default;
  }
});
Object.defineProperty(exports, "zhHK", {
  enumerable: true,
  get: function () {
    return _zhHK.default;
  }
});
Object.defineProperty(exports, "zhTW", {
  enumerable: true,
  get: function () {
    return _zhTW.default;
  }
});
var _deDE = _interopRequireDefault(require("./locales/de-DE"));
var _enUS = _interopRequireDefault(require("./locales/en-US"));
var _esES = _interopRequireDefault(require("./locales/es-ES"));
var _frFR = _interopRequireDefault(require("./locales/fr-FR"));
var _idID = _interopRequireDefault(require("./locales/id-ID"));
var _itIT = _interopRequireDefault(require("./locales/it-IT"));
var _jaJP = _interopRequireDefault(require("./locales/ja-JP"));
var _koKR = _interopRequireDefault(require("./locales/ko-KR"));
var _ptBR = _interopRequireDefault(require("./locales/pt-BR"));
var _ruRU = _interopRequireDefault(require("./locales/ru-RU"));
var _thTH = _interopRequireDefault(require("./locales/th-TH"));
var _viVN = _interopRequireDefault(require("./locales/vi-VN"));
var _zhCN = _interopRequireDefault(require("./locales/zh-CN"));
var _zhHK = _interopRequireDefault(require("./locales/zh-HK"));
var _zhTW = _interopRequireDefault(require("./locales/zh-TW"));
var _types = require("./locales/types");
Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _types[key];
    }
  });
});
// autogen by ../../../scripts/i18n/download.js

const defaultLocale = exports.defaultLocale = _enUS.default;
