// @universe-design/react/lib/shared/slot.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.renderSlotNode = exports.SLOT_RENDERER_SYMBOL = exports.SLOT_ELEMENT_TYPE_SYMBOL = void 0;
exports.resolveShorthand = resolveShorthand;
exports.resolveShorthandOld = resolveShorthandOld;
exports.resolveSlot = void 0;
var _react = _interopRequireDefault(require("react"));
var _react2 = require("@universe-design/stylex/react");
/* eslint-disable @typescript-eslint/ban-types */
/* eslint-disable @typescript-eslint/consistent-type-definitions */

function resolveShorthandOld(value) {
  if (typeof value === 'string' || typeof value === 'number' || Array.isArray(value) ||
  /*#__PURE__*/
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  _react.default.isValidElement(value)) {
    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
    return {
      children: value
    };
  }
  return value;
}
function resolveShorthand(value) {
  if (value === undefined) {
    return {};
  }
  if (typeof value === 'string' || typeof value === 'number' || Array.isArray(value) ||
  /*#__PURE__*/
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  _react.default.isValidElement(value)) {
    return {
      children: value
    };
  }
  return value;
}

// 添加 Slot 类型定义

const SLOT_ELEMENT_TYPE_SYMBOL = exports.SLOT_ELEMENT_TYPE_SYMBOL = Symbol.for('ud.slotElementType');
const SLOT_RENDERER_SYMBOL = exports.SLOT_RENDERER_SYMBOL = Symbol.for('ud.slotRenderer');

/**
 * 定义 `SlotNode` 接口，用于表示解析后的插槽节点。
 * 该接口包含插槽的属性、组件类型以及可选的插槽渲染函数。
 *
 * @typeParam Props - 插槽组件的属性类型，必须扩展自 `UnknownSlotProps`。
 */

/**
 * 定义解析插槽时使用的选项类型。
 * 该类型包含解析插槽所需的元素类型和可选的默认属性。
 *
 * @typeParam Props - 插槽组件的属性类型，必须扩展自 `UnknownSlotProps`。
 */

/**
 * 解析给定的插槽属性值，将其转换为 `SlotNode` 对象。
 * 如果传入的插槽属性值为 `null`，则返回 `null`。
 *
 * @typeParam Props - 插槽组件的属性类型，必须扩展自 `UnknownSlotProps`。
 * @param propSlot - 要解析的插槽属性值，类型为 `Slot<Props>`。
 * @param options - 解析插槽时使用的选项，包含元素类型和默认属性。
 * @returns 解析后的 `SlotNode<Props>` 对象，如果 `propSlot` 为 `null` 则返回 `null`。
 */
const resolveSlot = (propSlot, options) => {
  const {
    defaultProps,
    elementType,
    renderByDefault = false
  } = options;
  if (propSlot === null || propSlot === undefined && !renderByDefault) {
    return null;
  }
  const slot = resolveShorthand(propSlot);
  const {
    slotRenderer,
    ...componentProps
  } = slot;
  return {
    props: {
      ...defaultProps,
      ...componentProps
    },
    [SLOT_ELEMENT_TYPE_SYMBOL]: elementType,
    [SLOT_RENDERER_SYMBOL]: slotRenderer
  };
};

/**
 * 根据传入的 `SlotNode` 对象渲染对应的 React 节点。
 * 如果 `SlotNode` 为 `null`，则返回 `null`。
 * 支持使用传入的 `children` 覆盖 `SlotNode` 中的 `children` 属性。
 * 如果 `SlotNode` 中存在 `slotRenderer`，则使用该渲染函数进行渲染；否则，使用 `React.createElement` 进行渲染。
 *
 * @typeParam Props - 插槽组件的属性类型，必须扩展自 `UnknownSlotProps`。
 * @param slotNode - 要渲染的 `SlotNode` 对象，可能为 `null`。
 * @param children - 可选的子节点，用于覆盖 `SlotNode` 中的 `children` 属性。仅当 `Props` 中包含 `children` 属性时可用。
 * @returns 渲染后的 React 节点，如果 `slotNode` 为 `null` 则返回 `null`。
 */
exports.resolveSlot = resolveSlot;
const renderSlotNode = (slotNode, children) => {
  if (slotNode === null) {
    return null;
  }
  const C = slotNode[SLOT_ELEMENT_TYPE_SYMBOL];
  const slotRenderer = slotNode[SLOT_RENDERER_SYMBOL];
  if (children) {
    slotNode.props.children = children;
  }
  if (slotRenderer) {
    return slotRenderer(C, slotNode.props);
  }
  return (0, _react2.jsx)(C, slotNode.props);
};
exports.renderSlotNode = renderSlotNode;
