// @universe-design/react/es/notification/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import React, { useRef } from 'react';
import ReactDOM from 'react-dom';
import { CrossRootConfigProvider } from '../config-provider';
import Notification, { createNotificationWrapper } from './Notification';
import './style';
let globalNotification;
function getDefaultContainer() {
  const div = document.createElement('div');
  document.body.appendChild(div);
  return div;
}
function singletonRef(elem) {
  if (elem) {
    globalNotification = elem;
  }
}

/**
 * 用于多次重复 renderGlobalToast 和销毁自动生成的 DOM 元素
 */
class NotificationWrapper extends React.Component {
  constructor() {
    super(...arguments);
    _defineProperty(this, "div", void 0);
    _defineProperty(this, "isDefaultContainer", true);
    _defineProperty(this, "getContainer", () => {
      const {
        getContainer
      } = this.props;
      if (!this.div) {
        const c = getContainer === null || getContainer === void 0 ? void 0 : getContainer();
        this.div = c || getDefaultContainer();
        if (c) {
          this.isDefaultContainer = false;
        }
      }
      return this.div;
    });
  }
  componentWillUnmount() {
    globalNotification = undefined;
    if (this.isDefaultContainer && this.div) {
      this.div.remove();
      this.div = undefined;
    }
  }
  render() {
    return /*#__PURE__*/React.createElement(Notification, _extends({}, this.props, {
      ref: singletonRef,
      getContainer: this.getContainer
    }));
  }
}
export function getGlobalNotification() {
  if (globalNotification) {
    return Promise.resolve(globalNotification);
  }
  return new Promise(resolve => {
    const div = getDefaultContainer();
    // 如果没有使用 useToast 或者 renderGlobalToast 初始化 toast，
    // 则默认取第一个 ConfigProvider 提供的配置
    ReactDOM.render(/*#__PURE__*/React.createElement(CrossRootConfigProvider, null, /*#__PURE__*/React.createElement(NotificationWrapper, {
      ref: () => {
        if (globalNotification) {
          resolve(globalNotification);
        }
      },
      getContainer: () => div
    })), div);
  });
}
export const notification = createNotificationWrapper(getGlobalNotification);
export { Notification };

/**
 * 用于创建可以获取 ConfigContext 的全局 Notification 单例
 * @example
 * ```tsx
 * <ConfigProvider>
 *   {renderGlobalNotification({ className: 'my-toast' }, document.getElementById('my-container'))}
 *   <App />
 * </ConfigProvider>
 * ```
 *
 * @param props: ToastProps
 * @param container?: HTMLElement
 * @returns React.ReactPortal
 */
export function renderGlobalNotification(props) {
  return /*#__PURE__*/React.createElement(NotificationWrapper, props);
}

/**
 * 获取一个局部的
 * @returns
 */
export function useNotification(props, container) {
  const ref = useRef(null);
  const mountDivRef = useRef();
  const notificationIns = createNotificationWrapper(() => {
    if (ref.current) {
      return Promise.resolve(ref.current);
    }
    return getGlobalNotification();
  });
  const _getDefaultContainer = () => {
    if (mountDivRef.current) {
      return mountDivRef.current;
    }
    mountDivRef.current = getDefaultContainer();
    return mountDivRef.current;
  };
  return [notificationIns, /*#__PURE__*/React.createElement(Notification, _extends({
    key: "use-notification"
  }, props, {
    ref: ref,
    getContainer: () => container || _getDefaultContainer()
  }))];
}
