// @universe-design/react/es/notification/manager.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import React, { createRef, useEffect, useRef } from 'react';
import ReactDOM from 'react-dom';
import clsx from 'clsx';
import { CSSTransition } from 'react-transition-group';
import { useContextSelector } from 'use-context-selector';
import merge from 'lodash/merge';
import { ConfigContext } from '../config-provider';
import { getCssScopeFormatter, useClassPrefix } from '../config-provider/hooks';
import TransitionGroup from './TransitionGroup';
import { useEventCallback } from '../hooks';
const DEFAULT_CLOSE_DURATION = 3000;
const DEFAULT_TRANSITION_DURATION = 300;
const DEFAULT_PLACEMENT = 'topRight';
export const BASE_MESSAGE_LENGTH = 20;
let messageIdentity = 0;
function getIdentity() {
  messageIdentity += 1;
  return `ud-msg-manager-${messageIdentity}`;
}
const ManagerWrapConsumer = props => {
  const managerGlobalRootPrefixCls = useClassPrefix();
  const managerGlobalToast = useContextSelector(ConfigContext, v => v.toast);
  return props.children({
    managerGlobalRootPrefixCls,
    managerGlobalToast
  });
};

/**
 * React.Children.map 会对 jsx 的 key 属性进行转码，参考：https://github.com/facebook/react/blob/c1a53ad2b2f29f30412c682b42bdd6ac3227eafa/packages/react/src/ReactChildren.js#L63
 * bota 不能支持 unicode 的转码，所以需要处理一下，参考 https://stackoverflow.com/a/30106551/3053930
 */
function getEscapedKey(key) {
  if (typeof key !== 'string') {
    return key;
  }
  const codeUnits = new Uint16Array(key.length);
  for (let i = 0; i < codeUnits.length; i++) {
    codeUnits[i] = key.charCodeAt(i);
  }
  return btoa(String.fromCharCode(...new Uint8Array(codeUnits.buffer))).replace(/=/g, '-');
}
export function isNotAutoClose(duration) {
  return duration === null || duration === 0;
}
function getDefaultMessages() {
  return {
    topRight: [],
    topLeft: [],
    bottomRight: [],
    bottomLeft: [],
    top: []
  };
}
function ManagerItem(props) {
  const {
    prefixCls,
    duration,
    startTimer: startTimerProp,
    stopTimer,
    children
  } = props;
  const ref = useRef(null);
  const startTimer = useEventCallback(startTimerProp);
  useEffect(() => {
    if (ref.current && !isNotAutoClose(duration)) {
      startTimer(duration);
    }
  }, [duration, startTimer]);
  return /*#__PURE__*/React.createElement("div", {
    ref: ref,
    className: `${prefixCls}-item`,
    onMouseEnter: stopTimer,
    onMouseLeave: () => startTimer()
  }, children);
}

/**
 * 包含五个方位的 managers
 */
class NotificationManager extends React.Component {
  constructor(props) {
    super(props);
    _defineProperty(this, "timers", {});
    _defineProperty(this, "containerRef", /*#__PURE__*/createRef());
    _defineProperty(this, "classPrefix", '');
    _defineProperty(this, "startTimer", (key, duration) => {
      if (isNotAutoClose(duration)) {
        return;
      }
      const timer = this.timers[key];
      if (timer) {
        window.clearTimeout(timer);
      }
      this.timers[key] = window.setTimeout(() => this.remove(key), duration !== null && duration !== void 0 ? duration : DEFAULT_CLOSE_DURATION);
      return timer;
    });
    _defineProperty(this, "stopTimer", key => {
      const timer = this.timers[key];
      if (timer) {
        window.clearTimeout(timer);
      }
      delete this.timers[key];
    });
    this.state = {
      destroied: false,
      messages: getDefaultMessages()
    };
  }

  /**
   * 全局 config
   */

  /**
   * 计算后的配置
   */
  get config() {
    return Object.assign({}, NotificationManager.config, this.props);
  }

  /**
   * 优先使用 add 方法的 config，其次是已经存在的 config，最后降级到 this.props 的配置
   * @param props
   * @param existIdx
   * @returns
   */
  getMessageConfig(props, existIdx, placement) {
    const key = props.key || getIdentity();
    const placementCfg = props.placement || placement;
    const {
      transitionProps,
      duration: durationProps
    } = this.config;
    const {
      messages
    } = this.state;
    const prevNotice = messages[placementCfg][existIdx];
    let duration;
    if (props.duration !== undefined) {
      duration = props.duration; // eslint-disable-line prefer-destructuring
    } else if ((prevNotice === null || prevNotice === void 0 ? void 0 : prevNotice.duration) !== undefined) {
      duration = prevNotice.duration; // eslint-disable-line prefer-destructuring,react/destructuring-assignment
    } else {
      duration = durationProps;
    }
    return {
      key,
      duration,
      placement,
      transitionProps: props.transitionProps || (prevNotice === null || prevNotice === void 0 ? void 0 : prevNotice.transitionProps) || transitionProps,
      onClose: props.onClose || (prevNotice === null || prevNotice === void 0 ? void 0 : prevNotice.onClose),
      renderMessage: props.renderMessage || (prevNotice === null || prevNotice === void 0 ? void 0 : prevNotice.renderMessage)
    };
  }
  updateTimers(msgs) {
    const {
      maxCount
    } = this.config;
    if (msgs) {
      // 隐藏的 message 关闭 timer，避免集体消失
      msgs.map((m, idx) => {
        if (!maxCount) {
          if (!this.timers[m.key]) {
            this.startTimer(m.key, m.duration);
          }
        } else if (maxCount > idx) {
          // 1. 首次插入的消息
          // 2. 首次从隐藏进入显示的消息
          if (!this.timers[m.key]) {
            this.startTimer(m.key, m.duration);
          }
        } else {
          this.stopTimer(m.key);
        }
        return null;
      });
    }
  }
  add(props) {
    const {
      messages
    } = this.state;
    const {
      placement = DEFAULT_PLACEMENT,
      maxCount
    } = this.config;
    const placementCfg = props.placement || placement;
    let existIdx = -1;
    if (props.key) {
      existIdx = messages[placementCfg].findIndex(m => m.key === props.key);
    }
    const config = this.getMessageConfig(props, existIdx, placementCfg);
    const {
      key
    } = config;
    const newMessages = messages[placementCfg].concat();
    if (existIdx > -1) {
      this.stopTimer(config.key);
      newMessages[existIdx] = config;
    } else {
      newMessages.push(config);
      maxCount && newMessages.length > maxCount && newMessages.shift();
    }
    messages[placementCfg] = newMessages;
    this.setState({
      messages
    }, () => this.updateTimers(newMessages));
    return key;
  }
  remove(key) {
    const {
      messages
    } = this.state;
    let foundMsg = null;
    const newMessages = getDefaultMessages();
    const placements = Object.keys(messages);
    for (let i = 0; i < placements.length; i++) {
      const placement = placements[i];
      const msgs = messages[placement];
      const msg = msgs === null || msgs === void 0 ? void 0 : msgs.find(m => m.key === key);
      if (msg) {
        foundMsg = msg;
        newMessages[placement] = msgs.filter(v => v.key !== key);
      } else {
        newMessages[placement] = msgs;
      }
    }
    if (foundMsg) {
      const {
        placement
      } = foundMsg;
      this.stopTimer(key);
      this.setState({
        messages: newMessages
      }, () => placement && this.updateTimers(newMessages[placement]));
    }
  }

  /**
   * 清除除了 container 以外的
   */
  destroy() {
    Object.keys(this.timers).forEach(key => this.stopTimer(key));
    this.setState({
      messages: getDefaultMessages(),
      destroied: true
    });
  }
  handleClose(message) {
    message.key && this.remove(message.key);
  }
  renderMessage(message, rootPrefixCls) {
    const {
      key,
      duration,
      renderMessage,
      onClose,
      transitionProps: messageTransitionProps
    } = message;
    const {
      transitionProps,
      className
    } = this.config;
    const transProps = {
      timeout: DEFAULT_TRANSITION_DURATION,
      ...transitionProps,
      ...messageTransitionProps
    };
    const startTimer = newDration => this.startTimer(key, newDration || duration);
    const stopTimer = () => this.stopTimer(key);
    return /*#__PURE__*/React.createElement(CSSTransition, _extends({}, transProps, {
      key: getEscapedKey(key),
      onEnter: node => {
        /**
         * 当重复调用 open/close 时，上次 exit 可能没完成
         * 这里清除一下上次残留的状态
         */
        node.style.height = '';
      },
      onExit: node => {
        node.style.height = `${node.scrollHeight}px`;
      },
      onExiting: node => {
        node.style.height = '0';
      },
      onExited: () => {
        onClose && onClose();
      }
    }), status => /*#__PURE__*/React.createElement(ManagerItem, {
      prefixCls: this.classPrefix,
      stopTimer: stopTimer,
      startTimer: startTimer,
      duration: duration
    }, renderMessage({
      close: () => this.handleClose(message),
      transitionStatus: status,
      startTimer,
      stopTimer,
      rootPrefixCls,
      toastConfigCls: className
    })));
  }
  renderManager(placement) {
    const {
      messages
    } = this.state;
    const msgs = messages[placement];
    const {
      className,
      style,
      top,
      bottom,
      prefixCls,
      getContainer
    } = this.config;
    const container = getContainer === null || getContainer === void 0 ? void 0 : getContainer();
    const mergeStyle = {
      ...style,
      top,
      bottom
    };
    return /*#__PURE__*/React.createElement(ManagerWrapConsumer, {
      key: placement
    }, toastGlobalConfig => {
      const {
        managerGlobalToast,
        managerGlobalRootPrefixCls
      } = toastGlobalConfig;
      if (!this.classPrefix) {
        this.classPrefix = getCssScopeFormatter(managerGlobalRootPrefixCls, 'msg-manager', prefixCls)();
      }
      const cls = clsx(className, this.classPrefix, `${this.classPrefix}--${placement}`, {
        [`${this.classPrefix}--sticky`]: !container
      });
      return /*#__PURE__*/React.createElement("div", {
        ref: this.containerRef,
        className: cls,
        style: `${placement}` === 'top' ? merge({
          top: managerGlobalToast === null || managerGlobalToast === void 0 ? void 0 : managerGlobalToast.top
        }, mergeStyle) : mergeStyle
      }, /*#__PURE__*/React.createElement(TransitionGroup, null, msgs.map(message => this.renderMessage(message, managerGlobalRootPrefixCls))));
    });
  }
  render() {
    const {
      messages = []
    } = this.state;
    const {
      getContainer
    } = this.config;
    const container = getContainer === null || getContainer === void 0 ? void 0 : getContainer();
    const managers = /*#__PURE__*/React.createElement(React.Fragment, null, Object.keys(messages).map(placement => this.renderManager(placement)));
    if (container) {
      return /*#__PURE__*/ReactDOM.createPortal(managers, container);
    }
    return managers;
  }
}
_defineProperty(NotificationManager, "config", {});
export default NotificationManager;
