// @universe-design/react/es/notification/Notification.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _extends from "@babel/runtime/helpers/esm/extends";
import React, { createRef } from 'react';
import pick from 'lodash/pick';
import omit from 'lodash/omit';
import { Notice } from './notice';
import NotificationManager from './manager';

/**
 * 所有命令式调用方法的第一个参数
 */

const ManagerProps = ['className', 'prefixCls', 'duration', 'transitionProps', 'maxCount', 'getContainer', 'top', 'bottom', 'placement'];

/**
 * @docgen
 */

/**
 * @docgen
 */

export function createNotificationWrapper(getInstance) {
  return {
    /** 打开通知 */
    open(config) {
      return getInstance().then(t => t.open(config));
    },
    /** 信息通知 */
    info(configOrContent, duration, onClose) {
      return getInstance().then(t => t.info(configOrContent, duration, onClose));
    },
    /** 成功通知 */
    success(configOrContent, duration, onClose) {
      return getInstance().then(t => t.success(configOrContent, duration, onClose));
    },
    /** 警告通知 */
    warning(configOrContent, duration, onClose) {
      return getInstance().then(t => t.warning(configOrContent, duration, onClose));
    },
    /** 同 `warning` */
    warn(configOrContent, duration, onClose) {
      return getInstance().then(t => t.warn(configOrContent, duration, onClose));
    },
    /** 错误通知 */
    error(configOrContent, duration, onClose) {
      return getInstance().then(t => t.error(configOrContent, duration, onClose));
    },
    /** 通知全局配置 */
    config(options) {
      getInstance().then(t => t.config(options));
    },
    /** 同 `destroy` */
    close(key) {
      getInstance().then(t => t.close(key));
    },
    /** 按 `key` 销毁通知，或销毁所有通知 */
    destroy(key) {
      getInstance().then(t => t.destroy(key));
    }
  };
}
function getManagerConfig(option) {
  const {
    key,
    duration,
    onClose,
    placement,
    ...rest
  } = option;
  return {
    key,
    // 如果自定义了操作按钮，则不自动关闭
    duration: rest.btn ? null : duration,
    placement,
    renderMessage: _ref => {
      let {
        close
      } = _ref;
      return /*#__PURE__*/React.createElement(Notice, _extends({}, rest, {
        onClose: close
      }));
    },
    onClose
  };
}
export default class Notification extends React.PureComponent {
  constructor() {
    super(...arguments);
    _defineProperty(this, "managerRef", /*#__PURE__*/createRef());
  }
  /**
   * 计算后的配置
   */
  get _config() {
    return Object.assign({}, Notification.globalConfig, this.props);
  }

  /* eslint-disable no-dupe-class-members */
  open(config) {
    const manager = this.managerRef.current;
    const handleClose = config.onClose;
    return new Promise(resolve => {
      config.onClose = () => {
        handleClose && handleClose();
        resolve(undefined);
      };
      (manager === null || manager === void 0 ? void 0 : manager.add(this._getManagerConfig(config))) || '';
    });
  }
  info(configOrContent, duration, onClose) {
    return this._open('info', configOrContent, duration, onClose);
  }
  success(configOrContent, duration, onClose) {
    return this._open('success', configOrContent, duration, onClose);
  }
  warning(configOrContent, duration, onClose) {
    return this._open('warning', configOrContent, duration, onClose);
  }
  warn(configOrContent, duration, onClose) {
    return this.warning(configOrContent, duration, onClose);
  }
  error(configOrContent, duration, onClose) {
    return this._open('error', configOrContent, duration, onClose);
  }
  config(props) {
    Notification.globalConfig = Object.assign(Notification.globalConfig, props);
    // Notification.config = props;
    this.forceUpdate();
  }
  destroy(key) {
    if (key) {
      var _this$managerRef$curr;
      (_this$managerRef$curr = this.managerRef.current) === null || _this$managerRef$curr === void 0 ? void 0 : _this$managerRef$curr.remove(key);
    } else {
      var _this$managerRef$curr2;
      (_this$managerRef$curr2 = this.managerRef.current) === null || _this$managerRef$curr2 === void 0 ? void 0 : _this$managerRef$curr2.destroy();
    }
  }
  close(key) {
    return this.destroy(key);
  }
  render() {
    return /*#__PURE__*/React.createElement(NotificationManager, _extends({
      ref: this.managerRef,
      transitionProps: {
        timeout: 400,
        classNames: 'notification-animation'
      },
      duration: 4000
    }, pick(this._config, ManagerProps)));
  }
  _getManagerConfig(config) {
    const props = Object.assign(omit(this.props, ManagerProps), omit(this._config, ManagerProps));
    return getManagerConfig({
      ...props,
      ...config
    });
  }
  _open(type, configOrContent, duration, onClose) {
    if (configOrContent && typeof configOrContent === 'object' && 'message' in configOrContent) {
      return this.open({
        ...configOrContent,
        type
      });
    } else {
      const d = typeof duration === 'function' ? duration() : duration;
      return this.open({
        type,
        message: configOrContent,
        duration: d,
        onClose
      });
    }
  }
  /* eslint-enable no-dupe-class-members */
}
/**
 * 全局 config
 */
_defineProperty(Notification, "globalConfig", {});
