// @universe-design/react/es/notification/notice.js
import React, { forwardRef, useRef, useEffect } from 'react';
import SucceedFilled from '@universe-design/icons-react/SucceedFilled';
import WarningFilled from '@universe-design/icons-react/WarningFilled';
import ErrorFilled from '@universe-design/icons-react/ErrorFilled';
import InfoFilled from '@universe-design/icons-react/InfoFilled';
import CloseOutlined from '@universe-design/icons-react/CloseOutlined';
import clsx from 'clsx';
import stylex from '@universe-design/stylex';
import { cssTypographyTitle4, cssTypographyBody0 } from '../stylex';
import { useCssScope } from '../config-provider/hooks';
import { Button } from '../button';
const typeIcons = {
  success: SucceedFilled,
  warning: WarningFilled,
  info: InfoFilled,
  error: ErrorFilled
};
export const Notice = /*#__PURE__*/forwardRef((_ref, ref) => {
  let {
    btn,
    prefixCls,
    className,
    closable = true,
    closeIcon,
    message,
    description,
    style,
    icon,
    type = 'info',
    content,
    onClick,
    onMouseEnter,
    onMouseLeave,
    onClose
  } = _ref;
  const defaultScope = useCssScope('notification');
  const cssScope = prefixCls || defaultScope();
  const callback = useRef(onClose);
  const cls = clsx(cssScope, className, {
    [`${cssScope}__notice`]: true,
    [`${cssScope}-closable`]: closable
  });
  useEffect(() => {
    callback.current = onClose;
  }, [onClose]);
  const onRequestClose = e => {
    onClose && onClose(e);
  };
  const renderIcon = React.useCallback(() => {
    if (icon) {
      return icon;
    } else if (type) {
      const Icon = typeIcons[type];
      return /*#__PURE__*/React.createElement(Icon, {
        className: `${cssScope}__notice-content-icon`
      });
    }
  }, [icon, type, cssScope]);
  return /*#__PURE__*/React.createElement("div", {
    className: cls,
    onClick: onClick,
    onMouseLeave: onMouseLeave,
    onMouseEnter: onMouseEnter,
    ref: ref
  }, content || /*#__PURE__*/React.createElement("div", {
    style: style,
    className: clsx(`${cssScope}__notice-content`, `${cssScope}__notice-${type}`)
  }, /*#__PURE__*/React.createElement("span", {
    className: `${cssScope}__notice-border`
  }), /*#__PURE__*/React.createElement("div", {
    className: `${cssScope}__notice-icon-${type ? type : 'default'}`
  }, renderIcon()), /*#__PURE__*/React.createElement("div", {
    className: `${cssScope}__notice-with-icon`
  }, /*#__PURE__*/React.createElement("div", {
    className: clsx(`${cssScope}__notice-title`, stylex(cssTypographyTitle4))
  }, message), /*#__PURE__*/React.createElement("div", {
    className: clsx(`${cssScope}__notice-des`, stylex(cssTypographyBody0))
  }, description)), btn && /*#__PURE__*/React.createElement("div", {
    className: `${cssScope}__notice-operation`
  }, btn), (() => {
    if (closable) {
      return /*#__PURE__*/React.createElement("span", {
        className: `${cssScope}__notice-close`,
        tabIndex: 0,
        onClick: onRequestClose
      }, closeIcon || /*#__PURE__*/React.createElement(Button, {
        className: `${cssScope}__notice-close-icon`,
        icon: /*#__PURE__*/React.createElement(CloseOutlined, null),
        type: "icon",
        size: "small"
      }));
    }
    return null;
  })()));
});
