// @universe-design/react/es/notification/TransitionGroup.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
/**
 * This file copy most code from react-transition-group
 */
import React, { isValidElement } from 'react';
import TransitionGroupContext from 'react-transition-group/esm/TransitionGroupContext';
import { getChildMapping, getInitialChildMapping, getNextChildMapping } from 'react-transition-group/esm/utils/ChildMapping';
function getElem(mapping, key) {
  return mapping && mapping[`.$${key}`];
}
function isElemIn(mapping, key) {
  const elem = getElem(mapping, key);
  return elem && elem.props && elem.props.in;
}

/**
 * // 删除中间元素
 * case1: [1, 2, 3] + [1, 3] => [1, 2, 3]
 * // 在前面插入
 * case2: [1, 2, 3] + [4, 1, 2, 3] => [4, 1, 2, 3]
 * // 在后面插入
 * case3: [1, 2, 3] + [1, 2, 3, 4] => [1, 2, 3, 4]
 * // 在前面插入，删除中间元素
 * case4: [1, 2, 3] + [4, 1, 3] => [4, 1, 2, 3]
 * // 删除第一个，并新增元素
 * case5: [1, 2, 3] + [2, 3, 4] => [1, 2, 3, 4]
 */
function mergeChildrenArray(prevChildrenKeys, prevMapping, nextChildren) {
  const next = nextChildren instanceof Array ? nextChildren : [nextChildren];
  const newChildrenSort = (prevChildrenKeys || []).filter(k => isElemIn(prevMapping, k));
  const firstPrevChildKey = newChildrenSort[0];
  // 判断 prevKeys 中的第一个 key 是否已经遍历过。
  // 如果 nextChildren 中删除了 prevKeys 中的第一个 key，则认为已经遍历过。
  let firstPassed = Boolean(next.find(c => {
    if (/*#__PURE__*/isValidElement(c)) {
      return c.key !== firstPrevChildKey;
    }
    return false;
  }));
  for (let i = 0; i < next.length; i++) {
    const child = next[i];
    if (!child || ! /*#__PURE__*/isValidElement(child) || !child.key) {
      continue;
    }
    if (newChildrenSort.find(k => k === child.key)) {
      if (firstPrevChildKey === child.key) {
        firstPassed = true;
      }
      continue;
    }
    if (firstPassed) {
      // prevKeys 中第一个 key 遍历过，则 push 到新数组
      newChildrenSort.push(child.key);
    } else {
      newChildrenSort.unshift(child.key);
    }
  }
  return newChildrenSort;
}
const defaultProps = {
  component: 'div',
  childFactory: child => child
};

/**
 * The `<TransitionGroup>` component manages a set of transition components
 * (`<Transition>` and `<CSSTransition>`) in a list. Like with the transition
 * components, `<TransitionGroup>` is a state machine for managing the mounting
 * and unmounting of components over time.
 *
 * Consider the example below. As items are removed or added to the TodoList the
 * `in` prop is toggled automatically by the `<TransitionGroup>`.
 *
 * Note that `<TransitionGroup>`  does not define any animation behavior!
 * Exactly _how_ a list item animates is up to the individual transition
 * component. This means you can mix and match animations across different list
 * items.
 */
class TransitionGroup extends React.Component {
  constructor(props) {
    super(props);

    // Initial children should all be entering, dependent on appear
    _defineProperty(this, "mounted", false);
    // node is `undefined` when user provided `nodeRef` prop
    _defineProperty(this, "handleExited", (child, node) => {
      const {
        children
      } = this.props;
      const currentChildMapping = getChildMapping(children);
      if (child.key && child.key in currentChildMapping) {
        return;
      }
      if (child.props.onExited) {
        child.props.onExited(node);
      }
      if (this.mounted) {
        this.setState(state => {
          const mapping = {
            ...state.children
          };
          if (child.key) {
            delete mapping[child.key];
          }
          return {
            children: mapping
          };
        });
      }
    });
    this.state = {
      contextValue: {
        isMounting: true
      },
      handleExited: this.handleExited,
      firstRender: true,
      childrenSort: [],
      children: {}
    };
  }
  static getDerivedStateFromProps(nextProps, _ref) {
    let {
      children: prevChildMapping,
      handleExited,
      firstRender,
      childrenSort
    } = _ref;
    return {
      childrenSort: mergeChildrenArray(childrenSort, prevChildMapping, nextProps.children),
      children: firstRender ? getInitialChildMapping(nextProps, handleExited) : getNextChildMapping(nextProps, prevChildMapping, handleExited),
      firstRender: false
    };
  }
  componentDidMount() {
    this.mounted = true;
    this.setState({
      contextValue: {
        isMounting: false
      }
    });
  }
  componentWillUnmount() {
    this.mounted = false;
  }
  render() {
    const {
      component: Component = defaultProps.component,
      childFactory = defaultProps.childFactory,
      ...props
    } = this.props;
    const {
      contextValue,
      childrenSort,
      children: mapping
    } = this.state;
    const children = childrenSort.map(key => {
      const elem = getElem(mapping, key);
      return childFactory(elem);
    });
    delete props.appear;
    delete props.enter;
    delete props.exit;
    if (Component === null) {
      return /*#__PURE__*/React.createElement(TransitionGroupContext.Provider, {
        value: contextValue
      }, children);
    }
    return /*#__PURE__*/React.createElement(TransitionGroupContext.Provider, {
      value: contextValue
    }, /*#__PURE__*/React.createElement(Component, props, children));
  }
}
export default TransitionGroup;
