// @universe-design/react/es/config-provider/hooks/index.js
import React from 'react';
import { useContextSelector, useContext } from 'use-context-selector';
import { ConfigContext } from '../ConfigContext';
import { InputGroupSizeContext } from '../size';
export function useComponentSize() {
  return useContextSelector(ConfigContext, o => o.componentSize);
}
export function useInputGroupSize() {
  return useContext(InputGroupSizeContext);
}
export function useClassPrefix() {
  return useContextSelector(ConfigContext, o => o.rootPrefixCls);
}
export function getCssScopeFormatter(classPrefix, scope, cuzScope) {
  return function () {
    let suffix = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
    return cuzScope ? `${cuzScope}${suffix}` : `${classPrefix}__${scope}${suffix}`;
  };
}
export function useCssScope(scope, cuzScope) {
  const classPrefix = useClassPrefix();
  return React.useMemo(() => getCssScopeFormatter(classPrefix, scope, cuzScope), [classPrefix, cuzScope, scope]);
}
export function useDirection() {
  return useContextSelector(ConfigContext, o => o.direction);
}
export function useGetTargetContainer() {
  return useContextSelector(ConfigContext, o => o.getTargetContainer);
}
export function useLocale() {
  return useContextSelector(ConfigContext, o => o.locale);
}
export function useLocaleReceiver(componentName) {
  const contextLocale = useLocale();
  return React.useMemo(() => contextLocale.components[componentName], [componentName, contextLocale.components]);
}
export function useTheme() {
  return useContextSelector(ConfigContext, o => o.theme);
}
