// @universe-design/styles/cjs/universe-design/index.dark.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.udtoken_colorpicker_orange = exports.udtoken_colorpicker_neutral = exports.udtoken_colorpicker_lime = exports.udtoken_colorpicker_indigo = exports.udtoken_colorpicker_green = exports.udtoken_colorpicker_carmine = exports.udtoken_colorpicker_blue = exports.udtoken_chatsender_stopgeneration_bg2_normal = exports.udtoken_chatsender_stopgeneration_bg1_pressed = exports.udtoken_chatsender_stopgeneration_bg1_normal = exports.udtoken_chatsender_stopgeneration_bg1_hover = exports.udtoken_btn_text_bg_pri_pressed = exports.udtoken_btn_text_bg_pri_hover = exports.udtoken_btn_text_bg_pri_focus = exports.udtoken_btn_text_bg_neutral_pressed = exports.udtoken_btn_text_bg_neutral_hover = exports.udtoken_btn_text_bg_neutral_focus = exports.udtoken_btn_text_bg_danger_pressed = exports.udtoken_btn_text_bg_danger_hover = exports.udtoken_btn_text_bg_danger_focus = exports.udtoken_btn_selected_text_success = exports.udtoken_btn_selected_text_danger = exports.udtoken_btn_selected_border_neutral = exports.udtoken_btn_selected_bg_success_press = exports.udtoken_btn_selected_bg_success_normal = exports.udtoken_btn_selected_bg_success_hover = exports.udtoken_btn_selected_bg_neutral_press = exports.udtoken_btn_selected_bg_neutral_normal = exports.udtoken_btn_selected_bg_neutral_hover = exports.udtoken_btn_selected_bg_danger_press = exports.udtoken_btn_selected_bg_danger_normal = exports.udtoken_btn_selected_bg_danger_hover = exports.udtoken_btn_se_bg_success_pressed = exports.udtoken_btn_se_bg_success_hover = exports.udtoken_btn_se_bg_success_focus = exports.udtoken_btn_se_bg_pri_pressed = exports.udtoken_btn_se_bg_pri_hover = exports.udtoken_btn_se_bg_pri_focus = exports.udtoken_btn_se_bg_neutral_pressed = exports.udtoken_btn_se_bg_neutral_hover = exports.udtoken_btn_se_bg_neutral_focus = exports.udtoken_btn_se_bg_danger_pressed = exports.udtoken_btn_se_bg_danger_hover = exports.udtoken_btn_se_bg_danger_focus = exports.udtoken_btn_pri_text_disabled = exports.udtoken_btn_ghost_bg = exports.udtoken_block_view_inline_bg_self = exports.udtoken_block_view_inline_bg_other = exports.udtoken_block_view_inline_bg_nopermission = exports.udtoken_AI_cot_shinytext_mask = void 0;
exports.udtoken_navigation_bar_tab_bg_black = exports.udtoken_navigation_bar_resize_handle = exports.udtoken_message_card_text_yellow = exports.udtoken_message_card_text_wathet = exports.udtoken_message_card_text_violet = exports.udtoken_message_card_text_turquoise = exports.udtoken_message_card_text_red = exports.udtoken_message_card_text_purple = exports.udtoken_message_card_text_orange = exports.udtoken_message_card_text_neutral = exports.udtoken_message_card_text_lime = exports.udtoken_message_card_text_indigo = exports.udtoken_message_card_text_green = exports.udtoken_message_card_text_carmine = exports.udtoken_message_card_text_blue = exports.udtoken_message_card_bg_yellow = exports.udtoken_message_card_bg_wathet = exports.udtoken_message_card_bg_violet = exports.udtoken_message_card_bg_turquoise = exports.udtoken_message_card_bg_red = exports.udtoken_message_card_bg_purple = exports.udtoken_message_card_bg_orange = exports.udtoken_message_card_bg_neutral = exports.udtoken_message_card_bg_mask_special = exports.udtoken_message_card_bg_mask_general = exports.udtoken_message_card_bg_lime = exports.udtoken_message_card_bg_light_neutral = exports.udtoken_message_card_bg_indigo = exports.udtoken_message_card_bg_green = exports.udtoken_message_card_bg_carmine = exports.udtoken_message_card_bg_blue = exports.udtoken_input_bg_disabled = exports.udtoken_imageview_tooltips_bg = exports.udtoken_imageview_toolbox_bg = exports.udtoken_imageview_icondisabled = exports.udtoken_imageview_bg_mask = exports.udtoken_dropdown_button_divider = exports.udtoken_component_text_disabled_loading = exports.udtoken_component_outlined_bg = exports.udtoken_component_bg_top_mask = exports.udtoken_component_bg_right_mask = exports.udtoken_component_bg_left_mask = exports.udtoken_component_bg_bottom_mask = exports.udtoken_colorpicker_yellow = exports.udtoken_colorpicker_wathet = exports.udtoken_colorpicker_violet = exports.udtoken_colorpicker_turquoise = exports.udtoken_colorpicker_sunflower = exports.udtoken_colorpicker_red = exports.udtoken_colorpicker_purple = void 0;
exports.udtoken_tag_bg_purple = exports.udtoken_tag_bg_orange_solid = exports.udtoken_tag_bg_orange_pressed = exports.udtoken_tag_bg_orange_hover = exports.udtoken_tag_bg_orange = exports.udtoken_tag_bg_lime_solid = exports.udtoken_tag_bg_lime_hover = exports.udtoken_tag_bg_lime = exports.udtoken_tag_bg_indigo_solid = exports.udtoken_tag_bg_indigo_hover = exports.udtoken_tag_bg_indigo = exports.udtoken_tag_bg_green_solid = exports.udtoken_tag_bg_green_hover = exports.udtoken_tag_bg_green = exports.udtoken_tag_bg_carmine_solid = exports.udtoken_tag_bg_carmine_hover = exports.udtoken_tag_bg_carmine = exports.udtoken_tag_bg_blue_solid = exports.udtoken_tag_bg_blue_pressed = exports.udtoken_tag_bg_blue_hover = exports.udtoken_tag_bg_blue_default = exports.udtoken_tag_bg_blue = exports.udtoken_tag_AI_trial_bg_text_normal = exports.udtoken_tag_AI_trial_bg_normal = exports.udtoken_tabs_view_bg_hover = exports.udtoken_tabs_line_border_card = exports.udtoken_table_bg_selected = exports.udtoken_table_bg_pressed = exports.udtoken_table_bg_hover = exports.udtoken_table_bg_head = exports.udtoken_table_bg_grey = exports.udtoken_table_bg_expand = exports.udtoken_tab_se_bg_unselected = exports.udtoken_tab_pri_text = exports.udtoken_tab_pri_bg = exports.udtoken_switch_handle_disabled = exports.udtoken_steps_bg_hover = exports.udtoken_sliding_block_bg_disabled_loading = exports.udtoken_skeleton_fg = exports.udtoken_skeleton_bg = exports.udtoken_side_navigation_fill_selected = exports.udtoken_side_navigation_fill_pressed = exports.udtoken_side_navigation_fill_hover = exports.udtoken_side_navigation_fill_focus = exports.udtoken_reaction_bg_grey_float = exports.udtoken_reaction_bg_grey = exports.udtoken_quote_bar_bg = exports.udtoken_progress_bg = exports.udtoken_navigation_bar_tab_bg_gray = exports.udtoken_navigation_bar_tab_bg_blue = void 0;
exports.udtoken_tag_text_lime = exports.udtoken_tag_text_indigo = exports.udtoken_tag_text_green = exports.udtoken_tag_text_carmine = exports.udtoken_tag_text_blue = exports.udtoken_tag_state_red_light = exports.udtoken_tag_state_purple_light = exports.udtoken_tag_state_orange_light = exports.udtoken_tag_state_grey_light = exports.udtoken_tag_state_green_light = exports.udtoken_tag_state_blue_light = exports.udtoken_tag_neutral_text_solid = exports.udtoken_tag_neutral_text_normal = exports.udtoken_tag_neutral_text_inverse_opacity = exports.udtoken_tag_neutral_text_inverse = exports.udtoken_tag_neutral_close_pressed = exports.udtoken_tag_neutral_close_normal = exports.udtoken_tag_neutral_close_hover = exports.udtoken_tag_neutral_close_disabled = exports.udtoken_tag_neutral_bg_solid = exports.udtoken_tag_neutral_bg_normal_pressed = exports.udtoken_tag_neutral_bg_normal_hover = exports.udtoken_tag_neutral_bg_normal = exports.udtoken_tag_neutral_bg_inverse_opacity = exports.udtoken_tag_neutral_bg_inverse = exports.udtoken_tag_close_orange_pressed = exports.udtoken_tag_close_orange_normal = exports.udtoken_tag_close_orange_hover = exports.udtoken_tag_close_orange_disabled = exports.udtoken_tag_close_blue_pressed = exports.udtoken_tag_close_blue_normal = exports.udtoken_tag_close_blue_hover = exports.udtoken_tag_close_blue_disabled = exports.udtoken_tag_bg_yellow_solid = exports.udtoken_tag_bg_yellow_hover = exports.udtoken_tag_bg_yellow = exports.udtoken_tag_bg_wathet_solid = exports.udtoken_tag_bg_wathet_hover = exports.udtoken_tag_bg_wathet = exports.udtoken_tag_bg_violet_solid = exports.udtoken_tag_bg_violet_hover = exports.udtoken_tag_bg_violet = exports.udtoken_tag_bg_turquoise_solid = exports.udtoken_tag_bg_turquoise_hover = exports.udtoken_tag_bg_turquoise = exports.udtoken_tag_bg_red_solid = exports.udtoken_tag_bg_red_hover = exports.udtoken_tag_bg_red = exports.udtoken_tag_bg_purple_solid = exports.udtoken_tag_bg_purple_hover = void 0;
exports.udtoken_upload_mask_img = exports.udtoken_upload_bg_error = exports.udtoken_tag_text_yellow = exports.udtoken_tag_text_wathet = exports.udtoken_tag_text_violet = exports.udtoken_tag_text_turquoise = exports.udtoken_tag_text_s_yellow = exports.udtoken_tag_text_s_wathet = exports.udtoken_tag_text_s_violet = exports.udtoken_tag_text_s_turquoise = exports.udtoken_tag_text_s_red = exports.udtoken_tag_text_s_purple = exports.udtoken_tag_text_s_orange = exports.udtoken_tag_text_s_lime = exports.udtoken_tag_text_s_indigo = exports.udtoken_tag_text_s_green = exports.udtoken_tag_text_s_carmine = exports.udtoken_tag_text_s_blue = exports.udtoken_tag_text_red = exports.udtoken_tag_text_purple = exports.udtoken_tag_text_orange = void 0;
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
exports.udtoken_AI_cot_shinytext_mask = 'rgba(0, 0, 0, 0.5)';
exports.udtoken_block_view_inline_bg_nopermission = 'rgba(235, 235, 235, 0.1)';
exports.udtoken_block_view_inline_bg_other = 'rgba(38, 85, 182, 0.35)';
exports.udtoken_block_view_inline_bg_self = '#75a4ff';
exports.udtoken_btn_ghost_bg = 'rgba(10, 10, 10, 0)';
exports.udtoken_btn_pri_text_disabled = 'rgba(235, 235, 235, 0.3)';
exports.udtoken_btn_se_bg_danger_focus = 'rgba(240, 91, 86, 0.2)';
exports.udtoken_btn_se_bg_danger_hover = 'rgba(240, 91, 86, 0.2)';
exports.udtoken_btn_se_bg_danger_pressed = 'rgba(240, 91, 86, 0.3)';
exports.udtoken_btn_se_bg_neutral_focus = 'rgba(235, 235, 235, 0.1)';
exports.udtoken_btn_se_bg_neutral_hover = 'rgba(235, 235, 235, 0.1)';
exports.udtoken_btn_se_bg_neutral_pressed = 'rgba(235, 235, 235, 0.2)';
exports.udtoken_btn_se_bg_pri_focus = 'rgba(76, 136, 255, 0.2)';
exports.udtoken_btn_se_bg_pri_hover = 'rgba(76, 136, 255, 0.2)';
exports.udtoken_btn_se_bg_pri_pressed = 'rgba(76, 136, 255, 0.3)';
exports.udtoken_btn_se_bg_success_focus = 'rgba(65, 158, 52, 0.2)';
exports.udtoken_btn_se_bg_success_hover = 'rgba(65, 158, 52, 0.2)';
exports.udtoken_btn_se_bg_success_pressed = 'rgba(65, 158, 52, 0.3)';
exports.udtoken_btn_selected_bg_danger_hover = 'rgba(240, 91, 86, 0.3)';
exports.udtoken_btn_selected_bg_danger_normal = 'rgba(240, 91, 86, 0.15)';
exports.udtoken_btn_selected_bg_danger_press = 'rgba(240, 91, 86, 0.3)';
exports.udtoken_btn_selected_bg_neutral_hover = 'rgba(235, 235, 235, 0.12)';
exports.udtoken_btn_selected_bg_neutral_normal = 'rgba(235, 235, 235, 0.08)';
exports.udtoken_btn_selected_bg_neutral_press = 'rgba(235, 235, 235, 0.16)';
exports.udtoken_btn_selected_bg_success_hover = 'rgba(65, 158, 52, 0.2)';
exports.udtoken_btn_selected_bg_success_normal = 'rgba(65, 158, 52, 0.15)';
exports.udtoken_btn_selected_bg_success_press = 'rgba(65, 158, 52, 0.3)';
exports.udtoken_btn_selected_border_neutral = 'rgba(255, 255, 255, 0.5)';
exports.udtoken_btn_selected_text_danger = '#f05b56';
exports.udtoken_btn_selected_text_success = '#419e34';
exports.udtoken_btn_text_bg_danger_focus = 'rgba(240, 91, 86, 0.2)';
exports.udtoken_btn_text_bg_danger_hover = 'rgba(240, 91, 86, 0.2)';
exports.udtoken_btn_text_bg_danger_pressed = 'rgba(240, 91, 86, 0.3)';
exports.udtoken_btn_text_bg_neutral_focus = 'rgba(235, 235, 235, 0.1)';
exports.udtoken_btn_text_bg_neutral_hover = 'rgba(235, 235, 235, 0.1)';
exports.udtoken_btn_text_bg_neutral_pressed = 'rgba(235, 235, 235, 0.2)';
exports.udtoken_btn_text_bg_pri_focus = 'rgba(76, 136, 255, 0.2)';
exports.udtoken_btn_text_bg_pri_hover = 'rgba(76, 136, 255, 0.2)';
exports.udtoken_btn_text_bg_pri_pressed = 'rgba(76, 136, 255, 0.3)';
exports.udtoken_chatsender_stopgeneration_bg1_hover = '#505050';
exports.udtoken_chatsender_stopgeneration_bg1_normal = '#424242';
exports.udtoken_chatsender_stopgeneration_bg1_pressed = '#454545';
exports.udtoken_chatsender_stopgeneration_bg2_normal = '#f5f5f5';
exports.udtoken_colorpicker_blue = '#4c88ff';
exports.udtoken_colorpicker_carmine = '#c24a8e';
exports.udtoken_colorpicker_green = '#51ba43';
exports.udtoken_colorpicker_indigo = '#5e68e8';
exports.udtoken_colorpicker_lime = '#a6c313';
exports.udtoken_colorpicker_neutral = '#757575';
exports.udtoken_colorpicker_orange = '#f89e44';
exports.udtoken_colorpicker_purple = '#8c55ec';
exports.udtoken_colorpicker_red = '#f05b56';
exports.udtoken_colorpicker_sunflower = '#f7e663';
exports.udtoken_colorpicker_turquoise = '#17cfb5';
exports.udtoken_colorpicker_violet = '#d661d6';
exports.udtoken_colorpicker_wathet = '#51c3f0';
exports.udtoken_colorpicker_yellow = '#fbcb46';
exports.udtoken_component_bg_bottom_mask = 'linear-gradient(to bottom, rgba(26, 26, 26, 0) 0%, #1a1a1a 100%)';
exports.udtoken_component_bg_left_mask = 'linear-gradient(to left, rgba(26, 26, 26, 0) 0%, #1a1a1a 100%)';
exports.udtoken_component_bg_right_mask = 'linear-gradient(to right, rgba(26, 26, 26, 0) 0%, #1a1a1a 100%)';
exports.udtoken_component_bg_top_mask = 'linear-gradient(to top, rgba(26, 26, 26, 0) 0%, #1a1a1a 100%)';
exports.udtoken_component_outlined_bg = 'rgba(10, 10, 10, 0)';
exports.udtoken_component_text_disabled_loading = '#757575';
exports.udtoken_dropdown_button_divider = '#505050';
exports.udtoken_imageview_bg_mask = 'rgba(0, 0, 0, 0.8)';
exports.udtoken_imageview_icondisabled = 'rgba(255, 255, 255, 0.6)';
exports.udtoken_imageview_toolbox_bg = 'rgba(41, 41, 41, 0.8)';
exports.udtoken_imageview_tooltips_bg = '#505050';
exports.udtoken_input_bg_disabled = 'rgba(235, 235, 235, 0.12)';
exports.udtoken_message_card_bg_blue = '#173166';
exports.udtoken_message_card_bg_carmine = '#591c3f';
exports.udtoken_message_card_bg_green = '#173b12';
exports.udtoken_message_card_bg_indigo = '#2a2d69';
exports.udtoken_message_card_bg_light_neutral = '#373737';
exports.udtoken_message_card_bg_lime = '#303804';
exports.udtoken_message_card_bg_mask_general = 'rgba(10, 10, 10, 0.3)';
exports.udtoken_message_card_bg_mask_special = 'rgba(10, 10, 10, 0.3)';
exports.udtoken_message_card_bg_neutral = '#434343';
exports.udtoken_message_card_bg_orange = '#4a2b10';
exports.udtoken_message_card_bg_purple = '#3f2073';
exports.udtoken_message_card_bg_red = '#591f1d';
exports.udtoken_message_card_bg_turquoise = '#173b36';
exports.udtoken_message_card_bg_violet = '#541854';
exports.udtoken_message_card_bg_wathet = '#103647';
exports.udtoken_message_card_bg_yellow = '#473409';
exports.udtoken_message_card_text_blue = '#8fb4ff';
exports.udtoken_message_card_text_carmine = '#fc94cf';
exports.udtoken_message_card_text_green = '#69cc5c';
exports.udtoken_message_card_text_indigo = '#aaaff8';
exports.udtoken_message_card_text_lime = '#a6c313';
exports.udtoken_message_card_text_neutral = '#cfcfcf';
exports.udtoken_message_card_text_orange = '#f89e44';
exports.udtoken_message_card_text_purple = '#c5a3ff';
exports.udtoken_message_card_text_red = '#f89e9b';
exports.udtoken_message_card_text_turquoise = '#17cfb5';
exports.udtoken_message_card_text_violet = '#e99be9';
exports.udtoken_message_card_text_wathet = '#51c3f0';
exports.udtoken_message_card_text_yellow = '#fcd456';
exports.udtoken_navigation_bar_resize_handle = '#4599ff';
exports.udtoken_navigation_bar_tab_bg_black = '#0d0e12';
exports.udtoken_navigation_bar_tab_bg_blue = '#141c2a';
exports.udtoken_navigation_bar_tab_bg_gray = '#242424';
exports.udtoken_progress_bg = 'rgba(235, 235, 235, 0.2)';
exports.udtoken_quote_bar_bg = '#5f5f5f';
exports.udtoken_reaction_bg_grey = '#101010';
exports.udtoken_reaction_bg_grey_float = '#1f1f1f';
exports.udtoken_side_navigation_fill_focus = 'rgba(235, 235, 235, 0.12)';
exports.udtoken_side_navigation_fill_hover = 'rgba(235, 235, 235, 0.05)';
exports.udtoken_side_navigation_fill_pressed = 'rgba(235, 235, 235, 0.12)';
exports.udtoken_side_navigation_fill_selected = 'rgba(235, 235, 235, 0.1)';
exports.udtoken_skeleton_bg = 'rgba(240, 240, 240, 0.08)';
exports.udtoken_skeleton_fg = 'rgba(240, 240, 240, 0.05)';
exports.udtoken_sliding_block_bg_disabled_loading = '#757575';
exports.udtoken_steps_bg_hover = 'rgba(76, 136, 255, 0.15)';
exports.udtoken_switch_handle_disabled = 'rgba(255, 255, 255, 0.3)';
exports.udtoken_tab_pri_bg = '#194294';
exports.udtoken_tab_pri_text = '#4c88ff';
exports.udtoken_tab_se_bg_unselected = 'rgba(235, 235, 235, 0.2)';
exports.udtoken_table_bg_expand = '#0a0a0a';
exports.udtoken_table_bg_grey = '#292929';
exports.udtoken_table_bg_head = '#373737';
exports.udtoken_table_bg_hover = '#434343';
exports.udtoken_table_bg_pressed = '#5f5f5f';
exports.udtoken_table_bg_selected = '#173166';
exports.udtoken_tabs_line_border_card = '#434343';
exports.udtoken_tabs_view_bg_hover = 'rgba(235, 235, 235, 0.05)';
exports.udtoken_tag_AI_trial_bg_normal = 'radial-gradient(100.62% 141.42% at 100% 0%, rgba(124, 74, 212, 0.18) 0%, rgba(123, 131, 247, 0.18) 79.5%)';
exports.udtoken_tag_AI_trial_bg_text_normal = 'radial-gradient(58.07% 105.94% at 95% 28.12%, #a575fa 0%, #9499f7 100%)';
exports.udtoken_tag_bg_blue = 'rgba(76, 136, 255, 0.2)';
exports.udtoken_tag_bg_blue_default = 'rgba(117, 164, 255, 0.12)';
exports.udtoken_tag_bg_blue_hover = 'rgba(117, 164, 255, 0.3)';
exports.udtoken_tag_bg_blue_pressed = 'rgba(117, 164, 255, 0.2)';
exports.udtoken_tag_bg_blue_solid = '#173166';
exports.udtoken_tag_bg_carmine = 'rgba(219, 94, 164, 0.2)';
exports.udtoken_tag_bg_carmine_hover = 'rgba(219, 94, 164, 0.3)';
exports.udtoken_tag_bg_carmine_solid = '#591c3f';
exports.udtoken_tag_bg_green = 'rgba(65, 158, 52, 0.2)';
exports.udtoken_tag_bg_green_hover = 'rgba(65, 158, 52, 0.3)';
exports.udtoken_tag_bg_green_solid = '#173b12';
exports.udtoken_tag_bg_indigo = 'rgba(123, 131, 247, 0.2)';
exports.udtoken_tag_bg_indigo_hover = 'rgba(123, 131, 247, 0.3)';
exports.udtoken_tag_bg_indigo_solid = '#2a2d69';
exports.udtoken_tag_bg_lime = 'rgba(130, 153, 10, 0.2)';
exports.udtoken_tag_bg_lime_hover = 'rgba(130, 153, 10, 0.3)';
exports.udtoken_tag_bg_lime_solid = '#303804';
exports.udtoken_tag_bg_orange = 'rgba(219, 112, 24, 0.2)';
exports.udtoken_tag_bg_orange_hover = 'rgba(219, 112, 24, 0.3)';
exports.udtoken_tag_bg_orange_pressed = 'rgba(219, 112, 24, 0.35)';
exports.udtoken_tag_bg_orange_solid = '#4a2b10';
exports.udtoken_tag_bg_purple = 'rgba(165, 117, 250, 0.2)';
exports.udtoken_tag_bg_purple_hover = 'rgba(165, 117, 250, 0.3)';
exports.udtoken_tag_bg_purple_solid = '#3f2073';
exports.udtoken_tag_bg_red = 'rgba(209, 70, 66, 0.2)';
exports.udtoken_tag_bg_red_hover = 'rgba(209, 70, 66, 0.3)';
exports.udtoken_tag_bg_red_solid = '#591f1d';
exports.udtoken_tag_bg_turquoise = 'rgba(31, 161, 143, 0.2)';
exports.udtoken_tag_bg_turquoise_hover = 'rgba(31, 161, 143, 0.3)';
exports.udtoken_tag_bg_turquoise_solid = '#173b36';
exports.udtoken_tag_bg_violet = 'rgba(214, 97, 214, 0.2)';
exports.udtoken_tag_bg_violet_hover = 'rgba(214, 97, 214, 0.3)';
exports.udtoken_tag_bg_violet_solid = '#541854';
exports.udtoken_tag_bg_wathet = 'rgba(16, 153, 204, 0.2)';
exports.udtoken_tag_bg_wathet_hover = 'rgba(16, 153, 204, 0.3)';
exports.udtoken_tag_bg_wathet_solid = '#103647';
exports.udtoken_tag_bg_yellow = 'rgba(212, 155, 11, 0.15)';
exports.udtoken_tag_bg_yellow_hover = 'rgba(240, 182, 34, 0.3)';
exports.udtoken_tag_bg_yellow_solid = '#473409';
exports.udtoken_tag_close_blue_disabled = 'rgba(117, 164, 255, 0.8)';
exports.udtoken_tag_close_blue_hover = '#75a4ff';
exports.udtoken_tag_close_blue_normal = 'rgba(117, 164, 255, 0.6)';
exports.udtoken_tag_close_blue_pressed = 'rgba(117, 164, 255, 0.8)';
exports.udtoken_tag_close_orange_disabled = 'rgba(219, 112, 24, 0.3)';
exports.udtoken_tag_close_orange_hover = '#db7018';
exports.udtoken_tag_close_orange_normal = 'rgba(219, 112, 24, 0.6)';
exports.udtoken_tag_close_orange_pressed = 'rgba(219, 112, 24, 0.8)';
exports.udtoken_tag_neutral_bg_inverse = '#5f5f5f';
exports.udtoken_tag_neutral_bg_inverse_opacity = 'rgba(235, 235, 235, 0.4)';
exports.udtoken_tag_neutral_bg_normal = 'rgba(166, 166, 166, 0.2)';
exports.udtoken_tag_neutral_bg_normal_hover = 'rgba(235, 235, 235, 0.3)';
exports.udtoken_tag_neutral_bg_normal_pressed = 'rgba(235, 235, 235, 0.4)';
exports.udtoken_tag_neutral_bg_solid = '#373737';
exports.udtoken_tag_neutral_close_disabled = 'rgba(235, 235, 235, 0.3)';
exports.udtoken_tag_neutral_close_hover = '#ebebeb';
exports.udtoken_tag_neutral_close_normal = 'rgba(235, 235, 235, 0.6)';
exports.udtoken_tag_neutral_close_pressed = 'rgba(235, 235, 235, 0.8)';
exports.udtoken_tag_neutral_text_inverse = '#ffffff';
exports.udtoken_tag_neutral_text_inverse_opacity = '#ffffff';
exports.udtoken_tag_neutral_text_normal = '#a6a6a6';
exports.udtoken_tag_neutral_text_solid = '#a6a6a6';
exports.udtoken_tag_state_blue_light = '#4c88ff';
exports.udtoken_tag_state_green_light = '#419e34';
exports.udtoken_tag_state_grey_light = '#757575';
exports.udtoken_tag_state_orange_light = '#db7018';
exports.udtoken_tag_state_purple_light = '#a575fa';
exports.udtoken_tag_state_red_light = '#f05b56';
exports.udtoken_tag_text_blue = '#8fb4ff';
exports.udtoken_tag_text_carmine = '#fc94cf';
exports.udtoken_tag_text_green = '#69cc5c';
exports.udtoken_tag_text_indigo = '#aaaff8';
exports.udtoken_tag_text_lime = '#a6c313';
exports.udtoken_tag_text_orange = '#f89e44';
exports.udtoken_tag_text_purple = '#c5a3ff';
exports.udtoken_tag_text_red = '#f89e9b';
exports.udtoken_tag_text_s_blue = '#8fb4ff';
exports.udtoken_tag_text_s_carmine = '#fc94cf';
exports.udtoken_tag_text_s_green = '#69cc5c';
exports.udtoken_tag_text_s_indigo = '#aaaff8';
exports.udtoken_tag_text_s_lime = '#a6c313';
exports.udtoken_tag_text_s_orange = '#f89e44';
exports.udtoken_tag_text_s_purple = '#c5a3ff';
exports.udtoken_tag_text_s_red = '#f89e9b';
exports.udtoken_tag_text_s_turquoise = '#17cfb5';
exports.udtoken_tag_text_s_violet = '#e99be9';
exports.udtoken_tag_text_s_wathet = '#51c3f0';
exports.udtoken_tag_text_s_yellow = '#fcd456';
exports.udtoken_tag_text_turquoise = '#17cfb5';
exports.udtoken_tag_text_violet = '#e99be9';
exports.udtoken_tag_text_wathet = '#51c3f0';
exports.udtoken_tag_text_yellow = '#fcd456';
exports.udtoken_upload_bg_error = 'rgba(235, 235, 235, 0.12)';
exports.udtoken_upload_mask_img = 'rgba(0, 0, 0, 0.7)';
