// @universe-design/react/es/button/ButtonBase.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React, { forwardRef, useCallback, useMemo, useRef } from 'react';
import clsx from 'clsx';
import isNil from 'lodash/isNil';
import { SpinLoadingSvg } from '../_assets/svgs';
import { useComponentSize, useCssScope } from '../config-provider/hooks';
import { getSizeAbbr, deprecatedWarning, warning } from '../util';
import composeRefs from '../util/composeRefs';
import { useLoading } from '../loading/useLoading';
import './style';
import { markAsDisabledableComp } from '../shared/markAsDisabledableComp';

// type ReactButtonProps = React.DetailedHTMLProps<
//   React.ButtonHTMLAttributes<HTMLButtonElement>,
//   HTMLButtonElement
// >;

/**
 * Deprecated 组件的分类：https://bytedance.feishu.cn/docs/doccngkxeLeLStbebeToeKkPW7b#
 * Breaking change
 * - 重写 type，type 表示按钮类型，每种类型有自己的限制条件，如:
 *  - filled 限制背景改变，color 不变
 *  - outlined 限制背不变
 *  - ghost 限制背景透明
 * - 重写样式，降低样式的层叠优先级
 * - 修正 ButtonLink, LinkButton 的实现
 * - 去掉部分 Button 负 margin
 * - 去掉 info danger 属性，统一使用 color
 *  - info => color="primary"
 *  - danger => color="danger"
 * - 去掉 ghost 属性，ghost => type="ghost"
 *
 */

/**
 * @public
 */

/**
 * 支持原生`button`的所有事件和属性。
 * @public
 * @docgen
 */

const customizeSizeSet = new Set(['icon']);
function getSizeClass(type) {
  return customizeSizeSet.has(type) ? `${type}-size` : 'size';
}
const ButtonCmp = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    prefixCls: customizePrefixCls,
    className,
    children,
    shape,
    htmlType = 'button',
    type = 'outlined',
    size: customizeSize,
    color = 'default',
    selected = false,
    loading: loadingProp = false,
    delay,
    icon,
    endIcon,
    block = false,
    noMinWidth = false,
    onClick,
    danger,
    info,
    ...rest
  } = props;
  process.env.NODE_ENV !== "production" ? warning(isNil(props.autoBlur), deprecatedWarning('button', 'autoBlur')) : void 0;
  const cssScope = useCssScope('button');
  const prefixCls = customizePrefixCls || cssScope();
  const [loading] = useLoading({
    loading: loadingProp,
    delay
  });
  const size = useComponentSize();
  const buttonRef = useRef();
  const mergedRef = useMemo(() => composeRefs(ref, buttonRef), [ref, buttonRef]);
  const sizeValue = getSizeAbbr(customizeSize || size || 'medium');
  let variant = type;
  if (type === 'primary') {
    variant = 'filled';
  }
  if (type === 'default') {
    variant = 'outlined';
  }
  let usedColor = color;
  if (usedColor === 'default' && danger) {
    usedColor = 'danger';
  }
  if (usedColor === 'default' && info) {
    usedColor = 'primary';
  }
  const classes = clsx(prefixCls, {
    [cssScope(`--${variant}`)]: variant,
    [cssScope(`--${variant}-${usedColor}`)]: usedColor,
    [cssScope(`--${variant}-${usedColor}-loading`)]: loading,
    [cssScope(`--${variant}-${usedColor}-selected`)]: selected,
    [cssScope('--loading')]: loading,
    [cssScope('--selected')]: selected,
    [cssScope(`--${shape}`)]: shape,
    [cssScope(`--${getSizeClass(variant)}-${sizeValue}`)]: sizeValue,
    [cssScope('--disabled')]: rest.disabled,
    [cssScope('--block')]: block,
    [cssScope('--autowidth')]: noMinWidth
  }, className);
  let leftIcon = icon;
  let rightIcon = endIcon;
  if (leftIcon && children) {
    leftIcon = /*#__PURE__*/React.createElement("span", {
      className: clsx(cssScope('__icon-inline'), cssScope('__icon-inline-start'))
    }, leftIcon);
  }
  if (rightIcon && children) {
    rightIcon = /*#__PURE__*/React.createElement("span", {
      className: clsx(cssScope('__icon-inline'), cssScope('__icon-inline-end'))
    }, rightIcon);
  }
  if (loading && !rest.disabled) {
    leftIcon = /*#__PURE__*/React.createElement("span", {
      className: clsx(cssScope('__loading-icon'), cssScope('__icon-inline'), children && cssScope('__icon-inline-start'))
    }, /*#__PURE__*/React.createElement(SpinLoadingSvg, null));
  }
  const handleClick = useCallback(e => {
    if (loading || rest.disabled) {
      e.stopPropagation();
      e.preventDefault();
    } else {
      onClick && onClick(e);
    }
  }, [loading, rest.disabled, onClick]);
  if (rest.href !== undefined) {
    return /*#__PURE__*/React.createElement("a", _extends({}, rest, {
      ref: mergedRef,
      onClick: handleClick,
      className: classes
      // disabled={disabled}
    }), leftIcon, children, rightIcon);
  }
  return /*#__PURE__*/ /* eslint-disable react/button-has-type */React.createElement("button", _extends({}, rest, {
    type: htmlType,
    ref: mergedRef,
    onClick: handleClick,
    className: classes
  }), leftIcon, children, rightIcon)
  /* eslint-enable react/button-has-type */;
});
const ButtonBase = /* #__PURE__*/markAsDisabledableComp(ButtonCmp);
ButtonBase.__UD_BUTTON = true;
if (process.env.NODE_ENV !== "production") {
  ButtonBase.displayName = 'ButtonBase';
}
export { ButtonBase };
