// @universe-design/react/es/button/Link.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React, { forwardRef, useImperativeHandle, useRef, useCallback } from 'react';
import clsx from 'clsx';
import { SpinLoadingSvg } from '../_assets/svgs';
import { useCssScope } from '../config-provider/hooks';
import { markAsDisabledableComp } from '../shared/markAsDisabledableComp';

/**
 * Deprecated
 * 这是一个根据 Universse Design 设计，做的一个 Button 抽象，它强大而灵活
 */

const LinkCmp = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    prefixCls: customizePrefixCls,
    className,
    children,
    loading = false,
    disabled = false,
    icon: iconProp,
    onClick,
    ...rest
  } = props;
  const cssScope = useCssScope('link');
  const prefixCls = customizePrefixCls || cssScope();
  const aRef = useRef();
  useImperativeHandle(ref, () => aRef.current);
  const classes = clsx(prefixCls, {
    [cssScope('--loading')]: loading && !disabled,
    [cssScope('--disabled')]: disabled
  }, className);
  let icon = iconProp;
  if (loading && !disabled) {
    icon = /*#__PURE__*/React.createElement(SpinLoadingSvg, null);
  }
  const handleClick = useCallback(e => {
    if (loading || disabled) {
      e.stopPropagation();
      e.preventDefault();
    } else {
      onClick && onClick(e);
    }
  }, [loading, disabled, onClick]);
  return /*#__PURE__*/React.createElement("a", _extends({}, rest, {
    ref: a => {
      if (a) {
        aRef.current = a;
      }
    },
    className: classes,
    onClick: handleClick
  }), icon, children);
});
const Link = /* #__PURE__*/markAsDisabledableComp(LinkCmp);
Link.__UD_LINK = true;
Link.__UD_BUTTON = true;
if (process.env.NODE_ENV !== "production") {
  Link.displayName = 'Link';
}
export { Link };
