// @universe-design/react/lib/grid/row.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Row = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _context = _interopRequireDefault(require("./context"));
var _hooks = require("../config-provider/hooks");
var _util = require("../util");
require("./style");
/**
 * @public
 */

const Row = exports.Row = /*#__PURE__*/React.forwardRef((props, ref) => {
  const {
    children,
    align,
    justify,
    gutter = 0,
    wrap,
    style,
    className,
    ...otherProps
  } = props;
  const cssScope = (0, _hooks.useCssScope)('row');
  const classes = (0, _clsx.default)(cssScope(), {
    [cssScope(`--align-${align}`)]: align,
    [cssScope(`--justify-${justify}`)]: justify,
    [cssScope('--no-wrap')]: wrap === false
  }, className);
  const [screens, setScreens] = React.useState({
    xs: true,
    sm: true,
    md: true,
    lg: true,
    xl: true,
    xxl: true
  });
  const gutterRef = React.useRef(gutter);
  React.useEffect(() => {
    const token = _util.ResponsiveObserve.subscribe(screen => {
      const currentGutter = gutterRef.current || 0;
      if (!Array.isArray(currentGutter) && typeof currentGutter === 'object' || Array.isArray(currentGutter) && (typeof currentGutter[0] === 'object' || typeof currentGutter[1] === 'object')) {
        setScreens(screen);
      }
    });
    return () => _util.ResponsiveObserve.unsubscribe(token);
  }, []);
  const gutters = (() => {
    const results = [0, 0];
    const normalizedGutter = Array.isArray(gutter) ? gutter : [gutter, 0];
    normalizedGutter.forEach((g, index) => {
      if (typeof g === 'object') {
        for (let i = 0; i < _util.responsiveArray.length; i++) {
          const breakpoint = _util.responsiveArray[i];
          if (screens[breakpoint] && g[breakpoint] !== undefined) {
            results[index] = g[breakpoint];
            break;
          }
        }
      } else {
        results[index] = g || 0;
      }
    });
    return results;
  })();
  const rowStyle = React.useMemo(() => {
    const result = {};
    const [left, top] = gutters;
    if (left) {
      result.marginLeft = -left / 2;
      result.marginRight = -left / 2;
    }
    if (top) {
      result.marginTop = -top / 2;
      result.marginBottom = top / 2;
    }
    return result;
  }, [gutters]);
  return /*#__PURE__*/React.createElement(_context.default.Provider, {
    value: {
      gutter: gutters,
      wrap
    }
  }, /*#__PURE__*/React.createElement("div", (0, _extends2.default)({
    ref: ref,
    style: {
      ...rowStyle,
      ...style
    },
    className: classes
  }, otherProps), children));
});
if (process.env.NODE_ENV !== "production") {
  Row.displayName = 'Row';
}
