// @universe-design/react/lib/shared/components/popper-sidecar/FloatingUIPopperSidecar.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FloatingUIPopperSidecar = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _popper = require("../popper");
var _util = require("../../../util");
var _shared = require("./shared");
var _ConfigReducedMotion = require("../../../config-provider/ConfigReducedMotion");
const FloatingUIPopperSidecar = exports.FloatingUIPopperSidecar = /*#__PURE__*/React.forwardRef(function (_ref, ref) {
  var _popupAlignProp$overf, _popupAlignProp$overf2, _popupAlign$overflow, _popupAlign$overflow2, _popupAlign$overflow3;
  let {
    // config props
    action,
    popupPlacement,
    showArrow,
    arrowHasStroke,
    popupMotion,
    defaultMainAxisOffsetWithoutArrow,
    // user props
    prefixCls,
    enableAnimation: enableAnimationProp,
    popupAlign: popupAlignProp,
    offset: offsetProp,
    flip: flipProp,
    shift: shiftProp,
    ...restProps
  } = _ref;
  /**
   * Deprecated Warning
   */
  process.env.NODE_ENV !== "production" ? (0, _util.warning)(!popupAlignProp, (0, _util.deprecatedWarning)('FloatingUIPopperSidecar', 'popupAlign')) : void 0;
  process.env.NODE_ENV !== "production" ? (0, _util.warning)(!(popupAlignProp !== null && popupAlignProp !== void 0 && popupAlignProp.offset), (0, _util.deprecatedWarning)('FloatingUIPopperSidecar', 'popupAlign.offset', 'offset')) : void 0;
  process.env.NODE_ENV !== "production" ? (0, _util.warning)((popupAlignProp === null || popupAlignProp === void 0 ? void 0 : (_popupAlignProp$overf = popupAlignProp.overflow) === null || _popupAlignProp$overf === void 0 ? void 0 : _popupAlignProp$overf.adjustX) === undefined, (0, _util.deprecatedWarning)('FloatingUIPopperSidecar', 'popupAlign.overflow.adjustX', ['flip', 'shift'])) : void 0;
  process.env.NODE_ENV !== "production" ? (0, _util.warning)((popupAlignProp === null || popupAlignProp === void 0 ? void 0 : (_popupAlignProp$overf2 = popupAlignProp.overflow) === null || _popupAlignProp$overf2 === void 0 ? void 0 : _popupAlignProp$overf2.adjustY) === undefined, (0, _util.deprecatedWarning)('FloatingUIPopperSidecar', 'popupAlign.overflow.adjustY', ['flip', 'shift'])) : void 0;
  const popupAlign = popupAlignProp !== null && popupAlignProp !== void 0 ? popupAlignProp : {};
  const alignOffset = popupAlign === null || popupAlign === void 0 ? void 0 : popupAlign.offset;
  const points = popupAlign === null || popupAlign === void 0 ? void 0 : popupAlign.points;
  const adjustX = popupAlign === null || popupAlign === void 0 ? void 0 : (_popupAlign$overflow = popupAlign.overflow) === null || _popupAlign$overflow === void 0 ? void 0 : _popupAlign$overflow.adjustX;
  const adjustY = popupAlign === null || popupAlign === void 0 ? void 0 : (_popupAlign$overflow2 = popupAlign.overflow) === null || _popupAlign$overflow2 === void 0 ? void 0 : _popupAlign$overflow2.adjustY;
  const alwaysByViewport = popupAlign === null || popupAlign === void 0 ? void 0 : (_popupAlign$overflow3 = popupAlign.overflow) === null || _popupAlign$overflow3 === void 0 ? void 0 : _popupAlign$overflow3.alwaysByViewport;
  const placement = (() => {
    if ((0, _util.isNonNullable)(points)) {
      const [popupPoint, referencePoint] = points;
      // left
      if (popupPoint === 'tr' && referencePoint === 'tl') {
        return 'left-start';
      }
      if (popupPoint === 'cr' && referencePoint === 'cl') {
        return 'left';
      }
      if (popupPoint === 'br' && referencePoint === 'bl') {
        return 'left-end';
      }
      // right
      if (popupPoint === 'tl' && referencePoint === 'tr') {
        return 'right-start';
      }
      if (popupPoint === 'cl' && referencePoint === 'cr') {
        return 'right';
      }
      if (popupPoint === 'bl' && referencePoint === 'br') {
        return 'right-end';
      }
      // top
      if (popupPoint === 'bl' && referencePoint === 'tl') {
        return 'top-start';
      }
      if (popupPoint === 'bc' && referencePoint === 'tc') {
        return 'top';
      }
      if (popupPoint === 'br' && referencePoint === 'tr') {
        return 'top-end';
      }
      // bottom
      if (popupPoint === 'tl' && referencePoint === 'bl') {
        return 'bottom-start';
      }
      if (popupPoint === 'tc' && referencePoint === 'bc') {
        return 'bottom';
      }
      if (popupPoint === 'tr' && referencePoint === 'br') {
        return 'bottom-end';
      }
    }
    return (0, _util.getFloatingUIPlacement)(popupPlacement);
  })();
  const isTop = placement.startsWith('top');
  const isBottom = placement.startsWith('bottom');
  const isVertical = isTop || isBottom;
  const flip = React.useMemo(() => {
    if ((0, _util.isNonNullable)(flipProp)) {
      return flipProp;
    }
    if ((0, _util.isNonNullable)(adjustX) || (0, _util.isNonNullable)(adjustY)) {
      const flipAxisOptions = isVertical ? {
        // https://floating-ui.com/docs/flip#mainaxis
        mainAxis: adjustY !== undefined ? Boolean(adjustY) : true,
        // 如果 vertical，主轴是 y，对应 adjustY
        crossAxis: adjustX !== undefined ? Boolean(adjustX) : true // crossAxis flip 代表的是 xx-start -> xx-end 或反之
      } : {
        mainAxis: adjustX !== undefined ? Boolean(adjustX) : true,
        crossAxis: adjustY !== undefined ? Boolean(adjustY) : true
      };
      return flipAxisOptions;
    }
  }, [adjustX, adjustY, flipProp, isVertical]);
  const shift = React.useMemo(() => {
    if ((0, _util.isNonNullable)(shiftProp)) {
      return shiftProp;
    }
    if ((0, _util.isNonNullable)(adjustX) || (0, _util.isNonNullable)(adjustY)) {
      const shiftAxisOptions = isVertical ? {
        // https://floating-ui.com/docs/shift#mainaxis
        mainAxis: adjustX !== undefined ? Boolean(adjustX) : true // 如果 vertical，主轴是 x，我们不开启 crossAxis 的 shift（因为会盖住 reference），这种情况我们通过 flip + 内部滚动来解决
      } : {
        mainAxis: adjustY !== undefined ? Boolean(adjustY) : true
      };
      return shiftAxisOptions;
    }
  }, [adjustX, adjustY, isVertical, shiftProp]);
  const offset = React.useMemo(() => {
    if ((0, _util.isNonNullable)(alignOffset)) {
      const [x, y] = alignOffset;
      if (placement.startsWith('right')) {
        return {
          mainAxis: x,
          crossAxis: y
        };
      }
      if (placement.startsWith('left')) {
        return {
          mainAxis: -x,
          crossAxis: y
        };
      }
      if (placement.startsWith('bottom')) {
        return {
          mainAxis: y,
          crossAxis: x
        };
      }
      if (placement.startsWith('top')) {
        return {
          mainAxis: -y,
          crossAxis: x
        };
      }
    }
    if ((0, _util.isNonNullable)(offsetProp)) {
      return offsetProp;
    }
    return (showArrow ? _shared.POPUP_ARROW_SIZE.y : 0) + defaultMainAxisOffsetWithoutArrow;
  }, [alignOffset, defaultMainAxisOffsetWithoutArrow, offsetProp, placement, showArrow]);
  const enableAnimation = !(0, _ConfigReducedMotion.unstable_useConfigReducedMotion)((0, _util.isNonNullable)(enableAnimationProp) ? !enableAnimationProp : undefined);
  return /*#__PURE__*/React.createElement(_popper.Popper, (0, _extends2.default)({
    ref: ref,
    prefixCls,
    placement,
    trigger: action,
    arrow: showArrow ? /*#__PURE__*/React.createElement(_popper.PopperArrow, {
      hasStroke: arrowHasStroke
    }) : null,
    timeout: enableAnimation ? undefined : _popper.ZERO_TIMEOUT,
    offset,
    // 没有 arrow 的时候让 arrowPadding 是 0，这样可以跳过 arrow middleware，respect 安全边距而非箭头居中 https://github.com/floating-ui/floating-ui/pull/1496
    arrowPadding: showArrow ? _shared.POPUP_BORDER_RADIUS + _shared.POPUP_ARROW_SIZE.x : 0,
    flip,
    shift,
    rootBoundary: alwaysByViewport === false ? 'document' : undefined
  }, restProps));
});
if (process.env.NODE_ENV !== "production") {
  FloatingUIPopperSidecar.displayName = 'FloatingUIPopperSidecar';
}
