// @universe-design/react/lib/shared/components/popper-sidecar/RcTriggerPopperSidecar.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getPlacements = exports.RcTriggerPopperSidecar = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _trigger = require("../trigger");
var _StreamlineArrow = require("../trigger/StreamlineArrow");
var _shared = require("./shared");
var _util = require("../../../util");
var _ConfigReducedMotion = require("../../../config-provider/ConfigReducedMotion");
const getPlacements = _ref => {
  let {
    offset: offsetProp,
    defaultOffset
  } = _ref;
  let offset = offsetProp;
  if (offset === undefined || typeof offset !== 'number') {
    offset = defaultOffset;
  }
  const defaultAlign = {
    targetOffset: [0, 0],
    ignoreShake: true
  };
  const overflow = {
    adjustX: true,
    adjustY: true,
    alwaysByViewport: false
  };
  const left = {
    points: ['cr', 'cl'],
    overflow,
    offset: [-offset, 0],
    ...defaultAlign
  };
  const right = {
    points: ['cl', 'cr'],
    overflow,
    offset: [offset, 0],
    ...defaultAlign
  };
  const top = {
    points: ['bc', 'tc'],
    overflow,
    offset: [0, -offset],
    ...defaultAlign
  };
  const bottom = {
    points: ['tc', 'bc'],
    overflow,
    offset: [0, offset],
    ...defaultAlign
  };
  const topLeft = {
    points: ['bl', 'tl'],
    overflow,
    offset: [0, -offset],
    ...defaultAlign
  };
  const leftTop = {
    points: ['tr', 'tl'],
    overflow,
    offset: [-offset, 0],
    ...defaultAlign
  };
  const topRight = {
    points: ['br', 'tr'],
    overflow,
    offset: [0, -offset],
    ...defaultAlign
  };
  const rightTop = {
    points: ['tl', 'tr'],
    overflow,
    offset: [offset, 0],
    ...defaultAlign
  };
  const bottomRight = {
    points: ['tr', 'br'],
    overflow,
    offset: [0, offset],
    ...defaultAlign
  };
  const rightBottom = {
    points: ['bl', 'br'],
    overflow,
    offset: [offset, 0],
    ...defaultAlign
  };
  const bottomLeft = {
    points: ['tl', 'bl'],
    overflow,
    offset: [0, offset],
    ...defaultAlign
  };
  const leftBottom = {
    points: ['br', 'bl'],
    overflow,
    offset: [-offset, 0],
    ...defaultAlign
  };
  return {
    // legacy
    leftCenter: left,
    rightCenter: right,
    topCenter: top,
    bottomCenter: bottom,
    // in use
    left,
    right,
    top,
    bottom,
    topLeft,
    leftTop,
    topRight,
    rightTop,
    bottomRight,
    rightBottom,
    bottomLeft,
    leftBottom
  };
};
exports.getPlacements = getPlacements;
const RcTriggerPopperSidecar = exports.RcTriggerPopperSidecar = /*#__PURE__*/React.forwardRef(function (_ref2, ref) {
  let {
    // config props
    action,
    popupPlacement,
    showArrow,
    arrowHasStroke,
    popupMotion,
    defaultMainAxisOffsetWithoutArrow,
    // user props
    prefixCls,
    enableAnimation: enableAnimationProp,
    offset,
    // omit new popper props
    flip: flipProp,
    shift: shiftProp,
    ...restProps
  } = _ref2;
  /**
   * Deprecated Warning
   */
  process.env.NODE_ENV !== "production" ? (0, _util.warning)(false, (0, _util.devWarning)('RcTriggerPopperSidecar',
  // eslint-disable-next-line max-len
  'Detected that `RcTriggerPopperSidecar` is being used. `RcTriggerPopperSidecar` does not support follow scrolling, safe margin and other capabilities, and will be removed in the next major version, It is recommended to use `FloatingUIPopperSidecar` instead partialy. The usage is as follows `<ConfigProvider PopperSidecar={FloatingUIPopperSidecar} />`. Learn more about what we do here: https://bytedance.feishu.cn/wiki/wikcnG3DTElZpfwQqB88ArZQiKe')) : void 0;
  const builtinPlacements = React.useMemo(() => getPlacements({
    offset,
    defaultOffset: (showArrow ? _shared.POPUP_ARROW_SIZE.y : 0) + defaultMainAxisOffsetWithoutArrow
  }), [defaultMainAxisOffsetWithoutArrow, offset, showArrow]);
  const arrow = React.useMemo(() => showArrow ? {
    size: _shared.POPUP_ARROW_SIZE.x,
    safeMarginX: _shared.POPUP_BORDER_RADIUS,
    safeMarginY: _shared.POPUP_BORDER_RADIUS,
    content: _ref3 => {
      let {
        placement: arrowPlacement
      } = _ref3;
      const arrowStyle = (() => {
        if (arrowPlacement === 'top') {
          return {
            transform: 'rotate(180deg)',
            left: 0,
            bottom: 0
          };
        }
        if (arrowPlacement === 'bottom') {
          return {
            left: 0,
            top: 0
          };
        }
        if (arrowPlacement === 'left') {
          return {
            transform: 'rotate(180deg)',
            right: 0
          };
        }
        if (arrowPlacement === 'right') {
          return {
            left: 0
          };
        }
      })();
      return /*#__PURE__*/React.createElement(_StreamlineArrow.StreamlineArrow, {
        className: `${prefixCls}__arrow__streamline`,
        style: {
          position: 'absolute',
          ...arrowStyle
        },
        direction: ['left', 'right'].includes(arrowPlacement) ? 'vertical' : 'horizontal',
        hasStroke: arrowHasStroke
      });
    }
  } : undefined, [arrowHasStroke, prefixCls, showArrow]);
  const enableAnimation = !(0, _ConfigReducedMotion.unstable_useConfigReducedMotion)((0, _util.isNonNullable)(enableAnimationProp) ? !enableAnimationProp : undefined);
  return /*#__PURE__*/React.createElement(_trigger.UDTrigger, (0, _extends2.default)({
    ref: ref,
    prefixCls,
    popupPlacement,
    builtinPlacements,
    action,
    arrow,
    // 这里必须要提供一个 duration 0s 的 animation，因为 rc-motion 会根据 animationEnd 去判断 open 过程的结束，没有对应的 animation 会导致 popup 一直处于 pointer-events: none 导致用户无法移入 popup
    // https://code.byted.org/universe-ui/react/issues/982
    popupMotion: enableAnimation ? popupMotion : 'noop'
  }, restProps));
});
if (process.env.NODE_ENV !== "production") {
  RcTriggerPopperSidecar.displayName = 'RcTriggerPopperSidecar';
}
