// @universe-design/react/es/date-picker/panels/PanelHeaderSwitchers.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React from 'react';
import clsx from 'clsx';
import RightBoldOutlined from '@universe-design/icons-react/RightBoldOutlined';
import DownRightBoldOutlined from '@universe-design/icons-react/DownRightBoldOutlined';
import LeftBoldOutlined from '@universe-design/icons-react/LeftBoldOutlined';
import UpLeftBoldOutlined from '@universe-design/icons-react/UpLeftBoldOutlined';
import { Button } from '../../button';
import { useCssScope } from '../../config-provider';
export const usePickerHeaderSwitcherProps = _ref => {
  let {
    className,
    asPlaceholder = false,
    ...restProps
  } = _ref;
  const prefixCls = useCssScope('picker')();
  const headerIconCls = `${prefixCls}-panel-header-icon`;
  return {
    ...restProps,
    type: 'icon',
    size: 'extra-small',
    className: clsx(headerIconCls, asPlaceholder && `${headerIconCls}__hidden`, className)
  };
};
const NextMonthSwitcher = /*#__PURE__*/React.forwardRef((props, ref) => {
  const buttonProps = usePickerHeaderSwitcherProps(props);
  return /*#__PURE__*/React.createElement(Button, _extends({
    ref: ref
  }, buttonProps, {
    icon: /*#__PURE__*/React.createElement(RightBoldOutlined, null)
  }));
});
const NextYearSwitcher = /*#__PURE__*/React.forwardRef((props, ref) => {
  const buttonProps = usePickerHeaderSwitcherProps(props);
  return /*#__PURE__*/React.createElement(Button, _extends({
    ref: ref
  }, buttonProps, {
    icon: /*#__PURE__*/React.createElement(DownRightBoldOutlined, null)
  }));
});
const Next20YearsSwitcher = /*#__PURE__*/React.forwardRef((props, ref) => {
  const buttonProps = usePickerHeaderSwitcherProps(props);
  return /*#__PURE__*/React.createElement(Button, _extends({
    ref: ref
  }, buttonProps, {
    icon: /*#__PURE__*/React.createElement(DownRightBoldOutlined, null)
  }));
});
const PreMonthSwitcher = /*#__PURE__*/React.forwardRef((props, ref) => {
  const buttonProps = usePickerHeaderSwitcherProps(props);
  return /*#__PURE__*/React.createElement(Button, _extends({
    ref: ref
  }, buttonProps, {
    icon: /*#__PURE__*/React.createElement(LeftBoldOutlined, null)
  }));
});
const PreYearSwitcher = /*#__PURE__*/React.forwardRef((props, ref) => {
  const buttonProps = usePickerHeaderSwitcherProps(props);
  return /*#__PURE__*/React.createElement(Button, _extends({
    ref: ref
  }, buttonProps, {
    icon: /*#__PURE__*/React.createElement(UpLeftBoldOutlined, null)
  }));
});
const Pre20YearsSwitcher = /*#__PURE__*/React.forwardRef((props, ref) => {
  const buttonProps = usePickerHeaderSwitcherProps(props);
  return /*#__PURE__*/React.createElement(Button, _extends({
    ref: ref
  }, buttonProps, {
    icon: /*#__PURE__*/React.createElement(UpLeftBoldOutlined, null)
  }));
});
export { NextMonthSwitcher, NextYearSwitcher, Next20YearsSwitcher, PreMonthSwitcher, PreYearSwitcher, Pre20YearsSwitcher };
