// @universe-design/react/es/steps/Description.js
//.u19lj068{padding-right:8px;}
import u441rjq from "./../.stylex/u441rjq";
//.uxms9c{color:@icon-n3;}
import u1w9zb55 from "./../.stylex/u1w9zb55";
import stylex from '@universe-design/stylex';
import clsx from 'clsx';
import * as React from 'react';
import { cssTypographyBody2 } from '../stylex';
import { Text } from '../text';
import { useStepsContext } from './useSteps';
const sxDescription = {
  "props": {
    "sj55zd": u1w9zb55
  }
};
export default props => {
  var _steps$index;
  const {
    stepIndex
  } = props;
  const {
    steps,
    fixedDirection,
    cssScope,
    type,
    fixedLabelPlacement,
    showIcon,
    onStepClick,
    initial
  } = useStepsContext();
  const index = stepIndex - initial;
  const hasIcon = type !== 'navigation' || showIcon !== false;
  // eslint-disable-next-line @typescript-eslint/no-magic-numbers
  const iconSize = type === 'dot' ? 8 : hasIcon ? 28 : 0;
  const stemItemGap = index === 0 ? 0 : type === 'navigation' ? 24 : 8;
  const gapBetweenIconTitle = hasIcon ? 8 : 0;
  let margin = 0;
  if (fixedDirection === 'vertical') {
    margin = iconSize + gapBetweenIconTitle;
  } else {
    if (type === 'navigation' || fixedLabelPlacement === 'horizontal') {
      margin = iconSize + stemItemGap + gapBetweenIconTitle;
    } else {
      margin = 0;
    }
  }
  return /*#__PURE__*/React.createElement("div", {
    style: {
      paddingLeft: margin
    },
    onClick: () => {
      onStepClick(stepIndex);
    }
  }, /*#__PURE__*/React.createElement(Text, {
    as: "div",
    variant: "body0",
    key: index,
    sx: [cssTypographyBody2, sxDescription, fixedLabelPlacement === 'horizontal' && {
      "props": {
        "z189sj": u441rjq
      }
    }],
    className: clsx(cssScope('__item-description')),
    lineClamp: fixedDirection === 'horizontal' ? 2 : undefined
  }, (_steps$index = steps[index]) === null || _steps$index === void 0 ? void 0 : _steps$index.description));
};
