// @universe-design/react/es/steps/steps.js
//.u18eqifj{grid-template-columns:repeat(3, 1fr);}
import u1ltv0do from "./../.stylex/u1ltv0do";
import React, { forwardRef, cloneElement } from 'react';
import clsx from 'clsx';
import { isNonNullable, toFlatArray } from '../util';
import { useSteps, StepsContextProvider, useStepsContext } from './useSteps';
import { forwardHTMLRef } from '../util/forwardRef';
import NavigationIndicator from './NavigationIndicator';
import Description from './Description';
import stylex from '@universe-design/stylex';
const Descriptions = () => {
  const {
    steps,
    count,
    initial
  } = useStepsContext();
  const visibleSteps = steps.slice(0, count);
  return /*#__PURE__*/React.createElement(React.Fragment, null, visibleSteps.map((step, index) => /*#__PURE__*/React.createElement(Description, {
    stepIndex: index + initial,
    key: index
  })));
};
const StepsContent = forwardHTMLRef((props, ref) => {
  const {
    children
  } = props;
  const {
    style,
    cssScope,
    fixedDirection,
    activeIndex,
    initial,
    preStatus,
    fixedLabelPlacement,
    type,
    status,
    separateDescription,
    className,
    size
  } = useStepsContext();
  const childrenArray = toFlatArray(children);
  const childrenLength = childrenArray.length;
  const isDot = type === 'dot';
  const wrapperCls = clsx(cssScope(), cssScope(`--${fixedDirection}`), className, {
    [cssScope(`--${size}`)]: size,
    [cssScope('--navigation')]: type === 'navigation',
    [cssScope('--dot')]: isDot,
    [cssScope('--vertical-icon')]: fixedDirection === 'vertical' && !isDot,
    [cssScope(`--label-${fixedLabelPlacement}`)]: fixedDirection
  }, stylex({
    "props": {
      "Budl1dq": u1ltv0do
    }
  }));
  return /*#__PURE__*/React.createElement("div", {
    className: wrapperCls,
    ref: ref,
    style: {
      ...style,
      alignItems: fixedDirection === 'horizontal' && fixedLabelPlacement === 'horizontal' || type === 'navigation' ? 'center' : undefined,
      // columnGap: 8,
      gridTemplateColumns: fixedDirection === 'horizontal' ? `repeat(${childrenLength}, 1fr)` : '1fr'
    }
  }, childrenArray.filter(isNonNullable).map((child, index) => {
    if (typeof child === 'string' || typeof child === 'number' || typeof child === 'boolean' || !('props' in child)) {
      return null;
    }
    const stepNumber = initial + index;
    const childProps = {
      key: index,
      prefixCls: cssScope(),
      stepIndex: stepNumber,
      ...child.props
    };
    if (!child.props.status) {
      if (stepNumber < activeIndex) {
        childProps.status = 'finish';
      } else if (stepNumber === activeIndex) {
        childProps.status = status;
      } else {
        childProps.status = 'wait';
      }
    }
    const clx = clsx({
      [cssScope('--next-error')]: status === 'error' && index === activeIndex - 1,
      [cssScope(`--pre-${preStatus}`)]: preStatus !== null
    }, childProps.className);
    childProps.className = clx;

    // preStatus = childProps.status;

    return /*#__PURE__*/cloneElement(child, childProps);
  }), separateDescription && /*#__PURE__*/React.createElement(Descriptions, null), type === 'navigation' ? childrenArray.map((_, index) => /*#__PURE__*/React.createElement(NavigationIndicator, {
    index: index,
    key: index
  })) : null);
});
const Steps = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    children
  } = props;
  const stepsCtxValue = useSteps(props);
  return /*#__PURE__*/React.createElement(StepsContextProvider, {
    value: stepsCtxValue
  }, /*#__PURE__*/React.createElement(StepsContent, {
    ref: ref
  }, children));
});
if (process.env.NODE_ENV !== "production") {
  Steps.displayName = 'Steps';
}
export default Steps;
