// @universe-design/react/es/steps/step.js
//.u22iagw{display:flex;}
import ukchsq7 from "./../.stylex/ukchsq7";
//.u122n59{align-items:center;}
import ubyhozq from "./../.stylex/ubyhozq";
//.u1l02sjl{height:100%;}
import udg2fzc from "./../.stylex/udg2fzc";
/** @jsxRuntime classic */
/** @jsx jsx */
/* eslint-disable max-lines-per-function */

import React, { forwardRef } from 'react';
import clsx from 'clsx';
import DoneOutlined from '@universe-design/icons-react/DoneOutlined';
import CloseOutlined from '@universe-design/icons-react/CloseOutlined';
import RightBoldOutlined from '@universe-design/icons-react/RightBoldOutlined';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx } from '@universe-design/stylex/react';
import stylex from '@universe-design/stylex';
import { cssTypographyHeadline, cssTypographyBody0 } from '../stylex';
import { useStepsContext } from './useSteps';
import Description from './Description';
import { Text } from '../text';

/**
 * `<Step />` 组件的属性。
 * @public
 */

const Step = /*#__PURE__*/forwardRef((props, ref) => {
  var _steps;
  const {
    className,
    icon,
    title,
    subTitle,
    description,
    stepIndex = 0,
    status = 'wait',
    disabled
  } = props;
  const {
    registerNode,
    separateDescription,
    fixedDirection,
    clickable,
    steps,
    cssScope,
    activeIndex,
    showIcon,
    showNavIndicator,
    type,
    count,
    initial,
    progressDot,
    onStepClick
  } = useStepsContext();
  const prefixCls = cssScope();
  const index = stepIndex - initial;
  React.useEffect(() => {
    registerNode(stepIndex, {
      description,
      disabled,
      status
    });
  }, [description, disabled, registerNode, stepIndex, status]);
  const isLast = count - 1 === index;
  const preStatus = (_steps = steps[index - 1]) === null || _steps === void 0 ? void 0 : _steps.status;
  const wrapperCls = clsx(`${prefixCls}__item`, `${prefixCls}__item--${status}`, className, {
    [`${prefixCls}__item--last`]: isLast,
    [`${prefixCls}--pre-${preStatus}`]: preStatus,
    [`${prefixCls}__item--disabled`]: disabled,
    [`${prefixCls}__item--active`]: stepIndex === activeIndex,
    [`${prefixCls}__item--with-indicator`]: type === 'navigation' && showNavIndicator,
    [`${prefixCls}__item--without-icon`]: type === 'navigation' && !showIcon,
    [`${prefixCls}__item--clickable`]: clickable && !disabled
  }, ['process', 'error'].includes(status) ? stylex(cssTypographyHeadline) : stylex(cssTypographyBody0));
  function renderIconNode() {
    let iconNode = stepIndex + 1;
    if (type === 'dot') {
      iconNode = jsx("span", {
        className: `${prefixCls}__icon-dot`
      });
    } else if (icon) {
      iconNode = icon;
    } else if (status === 'finish') {
      iconNode = jsx(DoneOutlined, null);
    } else if (status === 'error') {
      iconNode = jsx(CloseOutlined, null);
    }
    const cls = clsx(`${prefixCls}__icon`, {
      [`${prefixCls}--include-icon`]: typeof iconNode !== 'number'
    });
    return jsx("span", {
      className: cls
    }, iconNode);
  }
  return jsx("div", {
    className: wrapperCls,
    ref: ref,
    sx: [separateDescription && {
      "props": {
        "mc9l5x": ukchsq7,
        "Bt984gj": ubyhozq,
        "Bqenvij": udg2fzc
      }
    }]
  }, jsx("div", {
    onClick: function () {
      onStepClick(stepIndex);
    },
    className: `${prefixCls}__item-container`
  }, jsx("div", {
    className: `${prefixCls}__item-content`
  }, !isLast && jsx("div", {
    className: `${prefixCls}__item-tail`
  }), function () {
    let iconWrapper;
    if (type === 'navigation' && !showIcon) {
      return null;
    }
    iconWrapper = jsx("div", {
      className: `${prefixCls}__item-icon`
    }, renderIconNode());
    if (typeof progressDot === 'function') {
      iconWrapper = progressDot(iconWrapper, {
        index: stepIndex,
        status,
        title,
        subTitle,
        description
      });
    }
    return iconWrapper;
  }(), jsx("div", {
    className: `${prefixCls}__item-title`
  }, jsx(Text, {
    className: `${prefixCls}__item-title-inner`,
    as: "div",
    lineClamp: fixedDirection === 'horizontal' ? 2 : undefined,
    variant: "plain"
  }, title, subTitle && jsx("span", {
    className: `${prefixCls}__item-subtitle`
  }, subTitle)))), !separateDescription && description && jsx(Description, {
    stepIndex: stepIndex
  }), !separateDescription && !isLast && jsx("div", {
    className: `${prefixCls}__item-padding`
  })), type === 'navigation' && !isLast && jsx("div", {
    className: `${prefixCls}__item-arrow`
  }, jsx(RightBoldOutlined, null)));
});
export default Step;
