// @universe-design/react/es/steps/useSteps.js
import React from 'react';
import { createContext, useContext, useContextSelector } from 'use-context-selector';
import { useImmer } from 'use-immer';
import { useCssScope } from '../config-provider';
import { createContextHook, useControlledState } from '@universe-design/react-utilities';
import { toFlatArray } from '../util';
export function useSteps(props) {
  const {
    className,
    style,
    initial = 0,
    current = 0,
    direction = 'horizontal',
    type = 'default',
    size,
    status = 'process',
    labelPlacement = 'vertical',
    progressDot = false,
    showNavIndicator = false,
    showIcon = false,
    onChange,
    children
  } = props;
  const cssScope = useCssScope('steps');

  // state
  const [steps, updateSteps] = useImmer([]);
  const [activeIndex, setActiveIndex] = useControlledState(0, current, {
    isControlledOverride: 'current' in props,
    onChange
  });
  // 这里用是否传了 onChange 作为是否可点击的判断依据
  const clickable = typeof onChange === 'function';

  /**
   * methods
   **/
  const registerNode = React.useCallback((index, stepInfo) => {
    updateSteps(draft => {
      draft[index] = stepInfo;
    });
  }, [updateSteps]);
  const onStepClick = React.useCallback(stepIndex => {
    if (!clickable) {
      return;
    }
    const stepDetail = steps[stepIndex - initial];
    if (!stepDetail.disabled && activeIndex !== stepIndex) {
      setActiveIndex(stepIndex);
    }
  }, [activeIndex, initial, setActiveIndex, steps, clickable]);
  let fixedLabelPlacement = labelPlacement;
  let fixedDirection = direction;
  if (direction === 'vertical') {
    fixedLabelPlacement = 'horizontal';
  }
  if (type === 'dot' && direction === 'horizontal') {
    fixedLabelPlacement = 'vertical';
  }
  if (type === 'navigation') {
    fixedDirection = 'horizontal';
    fixedLabelPlacement = showIcon ? 'horizontal' : 'vertical';
  }

  /**
   * 是否需要将 description 分离出来，单独布局
   */

  return {
    registerNode,
    steps,
    fixedDirection,
    fixedLabelPlacement,
    preStatus: null,
    showNavIndicator,
    progressDot,
    initial,
    current,
    activeIndex,
    setActiveIndex,
    status,
    size,
    style,
    className,
    cssScope,
    type,
    clickable: typeof onChange === 'function',
    showIcon,
    separateDescription: labelPlacement === 'horizontal' || type === 'navigation',
    count: toFlatArray(children).length,
    onStepClick
  };
}
const {
  Provider: StepsContextProvider,
  useContext: useStepsContext
} = createContextHook('StepsContext', {
  strict: true,
  customizedContextMethods: {
    createContext,
    useContext,
    useContextSelector
  }
});
export { StepsContextProvider, useStepsContext };
