// @universe-design/react/src/tag/Tag.tsx
import CloseBoldOutlined from '@universe-design/icons-react/CloseBoldOutlined';
import clsx from 'clsx';
import type { ReactElement } from 'react';
import React, { cloneElement, useRef } from 'react';
import stylex from '@universe-design/stylex';
import { useControlledState } from '@universe-design/react-utilities';
import {
  cssBlockBase,
  cssComponentBase,
  cssTypographyBody0,
  cssTypographyBody2,
  cssTypographyCaption0,
  cssTypographyCaption1Hack,
  cssTypographyHeadline,
  cssTypographyTitle4,
  cssTypographyTitle5,
  sxBorderRadius,
  sxOverflowEllipsis,
  sxTransitionTimingFunction,
} from '../stylex';
import { useCssScope } from '../config-provider/hooks';
import type { DefaultSizeType, DefaultExtraSizeType, LiteralUnion, ElementOf, StylexProps } from '../util';
import { devWarning, getSizeAbbr, isNonNullable, warning } from '../util';
import { isAvatar } from '../avatar/common';
import { useTextOverflowPopperSidecar } from '../config-provider/hooks/useTextOverflowPopperSidecar';
import { isHorizontalOverflowed } from '../util/dom';

import './style';

// xs
const tagSizeConfigXsFontSize = 10;
const tagSizeConfigXsCloseIconSize = 8;
const tagSizeConfigXsHeight = 16;
const tagSizeConfigXsPaddingX = 4;

// sm
const tagSizeConfigSmFontSize = 12;
const tagSizeConfigSmCloseIconSize = 10;
const tagSizeConfigSmHeight = 20;
const tagSizeConfigSmPaddingX = 6;

// md
const tagSizeConfigMdFontSize = 14;
const tagSizeConfigMdCloseIconSize = 12;
const tagSizeConfigMdHeight = 24;
const tagSizeConfigMdPaddingX = 6;

// lg
const tagSizeConfigLgFontSize = 16;
const tagSizeConfigLgCloseIconSize = 14;
const tagSizeConfigLgHeight = 32;
const tagSizeConfigLgPaddingX = 8;

const sxTag = {
  root: {
    base: stylex.compose(
      cssComponentBase,
      sxBorderRadius.xs,
      stylex.create({
        cursor: 'text',
        display: 'inline-flex',
        alignItems: 'center',
        whiteSpace: 'nowrap',
        transitionProperty: 'all',
        transitionDuration: '0.2s',
        maxWidth: '100%',
      }),
      sxTransitionTimingFunction.standard,
    ),
    disabled: stylex.create({
      cursor: 'not-allowed',
    }),
    xs: stylex.create({
      fontSize: tagSizeConfigXsFontSize,
      height: tagSizeConfigXsHeight,
      padding: `0 ${tagSizeConfigXsPaddingX}px`,
    }),
    sm: stylex.create({
      fontSize: tagSizeConfigSmFontSize,
      height: tagSizeConfigSmHeight,
      padding: `0 ${tagSizeConfigSmPaddingX}px`,
    }),
    md: stylex.create({
      fontSize: tagSizeConfigMdFontSize,
      height: tagSizeConfigMdHeight,
      padding: `0 ${tagSizeConfigMdPaddingX}px`,
    }),
    lg: stylex.create({
      fontSize: tagSizeConfigLgFontSize,
      height: tagSizeConfigLgHeight,
      padding: `0 ${tagSizeConfigLgPaddingX}px`,
    }),
    square: {
      base: stylex.create({
        padding: 0,
        justifyContent: 'center',
      }),
      xs: stylex.create({
        width: tagSizeConfigXsHeight,
      }),
      sm: stylex.create({
        width: tagSizeConfigSmHeight,
      }),
      md: stylex.create({
        width: tagSizeConfigMdHeight,
      }),
      lg: stylex.create({
        width: tagSizeConfigLgHeight,
      }),
    },
    round: {
      base: stylex.compose(
        stylex.create({
          padding: '0 8px',
        }),
        sxBorderRadius.xl,
      ),
      xs: stylex.create({
        paddingRight: tagSizeConfigXsHeight / 3,
      }),
      sm: stylex.create({
        paddingRight: tagSizeConfigSmHeight / 3,
      }),
      md: stylex.create({
        paddingRight: tagSizeConfigMdHeight / 3,
      }),
      lg: stylex.create({
        paddingRight: tagSizeConfigLgHeight / 3,
      }),
    },
  },
  statusIcon: {
    base: stylex.create({
      flexShrink: 0,
    }),
    xs: stylex.create({
      fontSize: 12,
    }),
  },
  content: {
    base: stylex.compose(cssBlockBase, sxOverflowEllipsis),
    disabled: stylex.create({
      opacity: 0.5,
    }),
    xs: cssTypographyCaption1Hack,
    sm: cssTypographyBody2,
    md: cssTypographyBody0,
    lg: cssTypographyTitle5,
    status: {
      xs: cssTypographyCaption1Hack,
      sm: cssTypographyCaption0,
      md: cssTypographyHeadline,
      lg: cssTypographyTitle4,
    },
    hasIcon: stylex.create({
      marginLeft: 4,
    }),
  },
  closeIcon: {
    base: stylex.compose(
      stylex.create({
        cursor: 'pointer',
        marginLeft: 4,
        opacity: 0.6,
        transitionProperty: 'all',
        transitionDuration: '0.2s',
        ':hover': {
          opacity: 1,
        },
        ':pressed': {
          opacity: 1,
        },
      }),
      sxTransitionTimingFunction.standard,
    ),
    disabled: stylex.create({
      opacity: 0.5,
      ':hover': {
        opacity: 0.5,
      },
      ':pressed': {
        opacity: 0.5,
      },
    }),
    xs: stylex.create({
      fontSize: tagSizeConfigXsCloseIconSize,
      marginLeft: 2,
    }),
    sm: stylex.create({
      fontSize: tagSizeConfigSmCloseIconSize,
    }),
    md: stylex.create({
      fontSize: tagSizeConfigMdCloseIconSize,
    }),
    lg: stylex.create({
      fontSize: tagSizeConfigLgCloseIconSize,
    }),
  },
};

function useTagCssScope() {
  return useCssScope('tag');
}

/**
 * 废弃的 color 命名，以及新命名
 */
const CompatibleColorMap = {
  'neutral-select': 'neutral-option',
  'neutral-light': 'neutral-solid',
} as const;

const TagPresetColorStatusEnum = [
  // Status Tag
  'waiting',
  'success',
  'info',
  'warning',
  'error',
  'missing',
] as const;

export const TagPresetColorEnum = [
  // Neutral
  'neutral',
  'neutral-solid',
  'neutral-option',
  'neutral-ghost',
  'neutral-ghost-transparent',
  // Attribute Tag
  'blue',
  'wathet',
  'turquoise',
  'green',
  'lime',
  'yellow',
  'orange',
  'red',
  'violet',
  'purple',
  'indigo',
  'carmine',
  // Solid Attribute Tag
  'blue-solid',
  'wathet-solid',
  'turquoise-solid',
  'green-solid',
  'lime-solid',
  'yellow-solid',
  'orange-solid',
  'red-solid',
  'violet-solid',
  'purple-solid',
  'indigo-solid',
  'carmine-solid',
  // Option Tag
  'blue-option',
  'wathet-option',
  'turquoise-option',
  'green-option',
  'lime-option',
  'yellow-option',
  'orange-option',
  'red-option',
  'violet-option',
  'purple-option',
  'indigo-option',
  'carmine-option',
  // Ghost Tag
  'blue-ghost',
  'carmine-ghost',
  'red-ghost',
  ...TagPresetColorStatusEnum,
  // Avatar Tag
  'neutral-avatar',
  'blue-avatar',
  'orange-avatar',
] as const;

/**
 * @public
 */
export type TagPropColor = LiteralUnion<ElementOf<typeof TagPresetColorEnum> | keyof typeof CompatibleColorMap, string>;

const getPresetColor = (color: TagPropColor): [isPresetColor: boolean, color: string] => {
  const compatibleColor = (CompatibleColorMap as Record<string, string>)[color];

  warning(
    Boolean(!compatibleColor),
    devWarning('Tag', `color value \`${color}\` is deprecated, please use \`${compatibleColor}\` instead.`),
  );

  const _color = compatibleColor ?? color;
  if ((TagPresetColorEnum as readonly string[]).includes(_color)) {
    return [true, _color];
  }
  return [false, _color];
};

export type TagPropSize = DefaultSizeType | Extract<DefaultExtraSizeType, 'extra-small'>;

/**
 * @public
 */
export interface TagProps extends React.HTMLAttributes<HTMLSpanElement>, StylexProps {
  /**
   * 标签颜色，包含预设颜色以及预设状态色，如果提供的颜色不在预设范围内，则会使用纯底色和 `@tag-other-color` 字色
   * @defaultValue `neutral`
   */
  color?: TagPropColor;
  /**
   * 标签的尺寸，拥有 `extra-small`、`small`、`medium` 和 `large` 四种尺寸
   * @defaultValue `medium`
   */
  size?: TagPropSize;
  /**
   * 标签形状，拥有 `round` 圆角矩形、`rect` 矩形和 `square` 方形，三种类型的形状
   * @defaultValue `rect`
   */
  shape?: 'round' | 'rect' | 'square';
  /**
   * 左侧状态图标
   */
  icon?: ReactElement;
  /**
   * 是否可关闭，`true` 则显示右侧可关闭图标，`false` 则不显示右侧可关闭图标。方形标签不能设置为可关闭。
   */
  closable?: boolean;
  /**
   * 是否显示标签的受控值
   */
  visible?: boolean;
  /**
   * 标签关闭的回调
   */
  onClose?: (e: React.MouseEvent<HTMLSpanElement>) => void;
  /**
   * 是否禁用
   * @defaultValue false
   */
  disabled?: boolean;
}

const tagAvatarSize = {
  'extra-small': 12,
  small: 16,
  medium: 20,
  large: 28,
};

function getAvatarSizeFromTagSize(size: NonNullable<TagProps['size']>) {
  return tagAvatarSize[size];
}

export const Tag = React.forwardRef<HTMLSpanElement, TagProps>((props, ref) => {
  const {
    color: colorProp = 'neutral',
    className,
    size: sizeProp = 'medium',
    icon,
    children,
    closable = false,
    visible: visibleProp,
    onClose,
    style,
    shape = 'rect',
    disabled = false,
    sx,
    ...restProps
  } = props;
  const size = getSizeAbbr(sizeProp);
  const cssScope = useTagCssScope();
  const TextOverflowPopper = useTextOverflowPopperSidecar();
  const [isPresetColor, color] = getPresetColor(colorProp);
  const tagStyle = {
    backgroundColor: color && !isPresetColor ? color : undefined,
    ...style,
  };
  const [visible, setVisible] = useControlledState(true, visibleProp);
  const onCloseClick = React.useCallback(
    (e: React.MouseEvent<HTMLSpanElement>) => {
      // 这个逻辑参考其他开源框架，不过感觉有受控 visible 不需要这个 prevented 判断
      e.stopPropagation();
      onClose?.(e);
      if (e.defaultPrevented) {
        return;
      }
      setVisible(false);
    },
    [onClose, setVisible],
  );

  const contentCls = clsx(cssScope('__content'), icon && cssScope('__content-has-icon'));
  const contentSx = [
    sxTag.content.base,
    disabled && sxTag.content.disabled,
    icon && sxTag.content.hasIcon,
    icon && size === 'xs' && stylex.create({ marginLeft: 2 }),
    size &&
      ((TagPresetColorStatusEnum as readonly string[]).includes(color)
        ? sxTag.content.status[size]
        : sxTag.content[size]),
  ];
  const contentRef = useRef<HTMLDivElement>(null);
  /** extra-small 因 zoom 问题, 暂时关闭 Tooltip */
  const contentNode = (() => {
    // square tag 没有地方放 children 了
    if (!isNonNullable(children) || shape === 'square') {
      return null;
    }

    return (
      <div ref={contentRef} className={clsx(contentCls, stylex(contentSx))}>
        {children}
      </div>
    );
  })();

  return visible ? (
    <TextOverflowPopper
      title={children}
      checkOverflow={() => {
        if (contentRef.current) {
          return isHorizontalOverflowed(contentRef.current);
        }
        return false;
      }}
    >
      <span
        ref={ref}
        className={clsx(
          cssScope(),
          className,
          size && cssScope(`-${size}`),
          shape && cssScope(`-${shape}`),
          cssScope(!isPresetColor ? '-other-color' : `-${color}`),
          icon && cssScope('-has-icon'),
          disabled && cssScope('-disabled'),
          stylex(
            sxTag.root.base,
            disabled && sxTag.root.disabled,
            size && sxTag.root[size],
            (shape === 'round' || shape === 'square') && sxTag.root[shape].base,
            size && (shape === 'round' || shape === 'square') && sxTag.root[shape][size],
            shape === 'round' &&
              icon &&
              stylex.create({
                paddingTop: 0,
                paddingLeft: 2,
                paddingBottom: 0,
              }),
            // 热区超出tag部分隐藏
            stylex.create({
              overflow: 'hidden',
            }),
            sx,
          ),
        )}
        style={tagStyle}
        {...restProps}
      >
        {icon
          ? cloneElement(icon as ReactElement, {
              className: clsx(
                cssScope('__status-icon'),
                icon?.props?.className,
                stylex(sxTag.statusIcon.base, size === 'xs' && sxTag.statusIcon[size]),
              ),
              ...(isAvatar(icon) ? { size: getAvatarSizeFromTagSize(sizeProp) } : {}),
            })
          : null}
        {contentNode}
        {/* square tag 没有地方放 close icon 了 */}
        {closable && shape !== 'square' ? (
          <CloseBoldOutlined
            className={clsx(
              cssScope('__close-icon'),
              stylex(
                sxTag.closeIcon.base,
                disabled && sxTag.closeIcon.disabled,
                size && sxTag.closeIcon[size],
                stylex.create({
                  position: 'relative',
                  '::after': {
                    display: 'block',
                    position: 'absolute',
                    width: 24,
                    height: 24,
                    content: '""',
                    top: 'calc(50% - 12px)', // (svg的高度 - 24) / 2
                    left: -2, // svg距离左侧最小间距
                  },
                }),
              ),
            )}
            onClick={onCloseClick}
          />
        ) : null}
      </span>
    </TextOverflowPopper>
  ) : null;
});
if (__DEV__) {
  Tag.displayName = 'Tag';
}
