// @universe-design/react/es/rate/index.js
import React, { useState, useRef, useEffect } from 'react';
import clsx from 'clsx';
import stylex from '@universe-design/stylex';
import { useControlledState } from '@universe-design/react-utilities';
import { cssTypographyBody0 } from '../stylex';
import { useCssScope, useLocale } from '../config-provider/hooks';
import { Tooltip } from '../tooltip';
import CollectFilled from '@universe-design/icons-react/CollectFilled';
import { isFunction } from '../util';
import noop from 'lodash/noop';
import './style';
export const Rate = props => {
  const {
    style,
    className,
    character,
    allowHalf,
    allowClear = true,
    value: valueInProps,
    defaultValue = 0,
    readonly = false,
    count = 5,
    tooltips = [],
    showDescriptions = true,
    descriptions = [],
    autoFocus = false,
    onHoverChange,
    onChange
  } = props;
  const cssScope = useCssScope('rate');
  const cls = clsx(cssScope(), className);
  const {
    components: {
      rate: rateLocale
    }
  } = useLocale();
  const rateRef = useRef(null);
  const [value, setValue] = useControlledState(defaultValue, valueInProps, {
    onChange
  });
  const [hoverValue, setHoverValue] = useState();
  const [cleanedValue, setCleanedValue] = useState();
  useEffect(() => {
    if (rateRef.current && autoFocus) {
      rateRef.current.focus();
    }
  }, [rateRef, autoFocus]);
  const getValue = (index, e) => {
    let currentValue = index + 1;
    const {
      left,
      width
    } = e.currentTarget.getBoundingClientRect();
    if (allowHalf && e.clientX - left < width / 2) {
      // eslint-disable-next-line @typescript-eslint/no-magic-numbers
      currentValue -= 0.5;
    }
    return currentValue;
  };
  const calculateDes = () => {
    const index = parseInt((hoverValue || value || 0).toFixed(), 10);
    return descriptions[index] || rateLocale.descriptions[index];
  };
  const handleMouseMove = (index, e) => {
    const currentValue = getValue(index, e);
    // 避免点击触发清除后，再次设置选中态
    /* istanbul ignore else  */
    if (currentValue !== cleanedValue) {
      setHoverValue(currentValue);
      isFunction(onHoverChange) && onHoverChange(currentValue);
    }
  };
  const handleChoose = (index, e) => {
    const currentValue = getValue(index, e);
    const needReset = allowClear && currentValue === value;
    setValue(needReset ? 0 : currentValue);
    setHoverValue(undefined);
    setCleanedValue(needReset ? currentValue : undefined);
  };
  const wrapTooltips = (index, node) => {
    const tip = tooltips[index];
    if (tip) {
      return /*#__PURE__*/React.createElement(Tooltip, {
        key: `rate-tooltip-${index}`,
        title: tip
      }, node);
    }
    return node;
  };
  const renderCharacterEl = index => {
    if (isFunction(character)) {
      return character({
        ...props,
        index
      });
    } else {
      return character || /*#__PURE__*/React.createElement(CollectFilled, null);
    }
  };
  const renderCharacter = index => {
    let preciseStyle;
    const currentIndex = index + 1;
    const currentValue = hoverValue || value;
    const includes = currentValue && currentValue > index && currentValue < currentIndex;
    const hasHalf = allowHalf ? includes : false;
    const itemCls = clsx(cssScope('__character'), {
      [cssScope('__character--full')]: currentValue && currentValue >= currentIndex,
      [cssScope('__character--half')]: readonly && includes || hasHalf
    });
    if (readonly && currentValue && includes) {
      preciseStyle = {
        width: `${(currentValue - index) * 100}%`
      };
    }
    return wrapTooltips(index, /*#__PURE__*/React.createElement("div", {
      className: itemCls,
      key: index,
      onMouseMove: readonly ? noop : e => handleMouseMove(index, e),
      onClick: readonly ? noop : e => handleChoose(index, e)
    }, /*#__PURE__*/React.createElement("div", {
      style: preciseStyle,
      className: cssScope('__character-half-body')
    }, /*#__PURE__*/React.createElement("div", {
      className: cssScope('__character-half-body-wrap')
    }, renderCharacterEl(index))), /*#__PURE__*/React.createElement("div", {
      className: cssScope('__character-body')
    }, renderCharacterEl(index))));
  };
  return /*#__PURE__*/React.createElement("div", {
    ref: rateRef,
    style: style,
    className: cls,
    onMouseLeave: () => {
      setHoverValue(undefined);
      setCleanedValue(undefined);
      isFunction(onHoverChange) && onHoverChange(0);
    }
  }, /*#__PURE__*/React.createElement("div", {
    className: cssScope('__characters')
  }, (() => {
    const items = [];
    for (let i = 0; i < count; i++) {
      items.push(renderCharacter(i));
    }
    return items;
  })()), (() => {
    if (showDescriptions) {
      return /*#__PURE__*/React.createElement("div", {
        className: clsx(cssScope('__des'), stylex(cssTypographyBody0))
      }, calculateDes());
    }
    return null;
  })());
};
