// @universe-design/react/lib/stylex/border-radius.sx.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.sxBorderRadius = void 0;
var _u1m59yg = _interopRequireDefault(require("./../.stylex/u1m59yg4"));
var _ubm82nm = _interopRequireDefault(require("./../.stylex/ubm82nm"));
var _u1qcfi6m = _interopRequireDefault(require("./../.stylex/u1qcfi6m"));
var _u1pc15gj = _interopRequireDefault(require("./../.stylex/u1pc15gj"));
var _uacbfjb = _interopRequireDefault(require("./../.stylex/uacbfjb"));
var _u1l86f2g = _interopRequireDefault(require("./../.stylex/u1l86f2g"));
var _u1r1mxth = _interopRequireDefault(require("./../.stylex/u1r1mxth"));
var _ul6wel = _interopRequireDefault(require("./../.stylex/ul6wel9"));
var _u1j8h7k = _interopRequireDefault(require("./../.stylex/u1j8h7k8"));
var _u10e4wt = _interopRequireDefault(require("./../.stylex/u10e4wt4"));
var _u5rwohh = _interopRequireDefault(require("./../.stylex/u5rwohh"));
var _u14kz1ee = _interopRequireDefault(require("./../.stylex/u14kz1ee"));
var _u1n69reo = _interopRequireDefault(require("./../.stylex/u1n69reo"));
var _u1sx1i = _interopRequireDefault(require("./../.stylex/u1sx1i35"));
var _ui28ht = _interopRequireDefault(require("./../.stylex/ui28ht5"));
var _uyycwk = _interopRequireDefault(require("./../.stylex/uyycwk0"));
var _uzvwtxa = _interopRequireDefault(require("./../.stylex/uzvwtxa"));
var _u5dwexa = _interopRequireDefault(require("./../.stylex/u5dwexa"));
var _ukikfbv = _interopRequireDefault(require("./../.stylex/ukikfbv"));
var _uhf7ou = _interopRequireDefault(require("./../.stylex/uhf7ou4"));
var _u1ubg0fh = _interopRequireDefault(require("./../.stylex/u1ubg0fh"));
var _u19y2ap = _interopRequireDefault(require("./../.stylex/u19y2ap9"));
var _u16wxydm = _interopRequireDefault(require("./../.stylex/u16wxydm"));
var _u5pet9v = _interopRequireDefault(require("./../.stylex/u5pet9v"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
//.u55qycf{border-top-left-radius:999999px;}
//.u2ie9lg{border-top-right-radius:999999px;}
//.u1ewtp5v{border-bottom-right-radius:999999px;}
//.u1j6dxmp{border-bottom-left-radius:999999px;}
//.ugv9nof{border-top-left-radius:12px;}
//.u24j1xx{border-top-right-radius:12px;}
//.u1ig3etr{border-bottom-right-radius:12px;}
//.uy0y5fd{border-bottom-left-radius:12px;}
//.umb70yw{border-top-left-radius:8px;}
//.u1spoy8{border-top-right-radius:8px;}
//.u1erghxr{border-bottom-right-radius:8px;}
//.u1ehz9de{border-bottom-left-radius:8px;}
//.uifjimi{border-top-left-radius:6px;}
//.u8rsvsd{border-top-right-radius:6px;}
//.u18fiblo{border-bottom-right-radius:6px;}
//.u1asnnh4{border-bottom-left-radius:6px;}
//.u8wkphi{border-top-left-radius:4px;}
//.uwsfkhu{border-top-right-radius:4px;}
//.uff7au0{border-bottom-right-radius:4px;}
//.u1bjk9e1{border-bottom-left-radius:4px;}
//.u1127022{border-top-left-radius:2px;}
//.u1stktl2{border-top-right-radius:2px;}
//.u2i1086{border-bottom-right-radius:2px;}
//.u1rlsnju{border-bottom-left-radius:2px;}
const sxBorderRadius = exports.sxBorderRadius = {
  xxs: {
    "props": {
      "Btl43ni": _u1ubg0fh.default,
      "B7oj6ja": _u19y2ap.default,
      "Bbmb7ep": _u16wxydm.default,
      "Beyfa6y": _u5pet9v.default
    }
  },
  xs: {
    "props": {
      "Btl43ni": _uzvwtxa.default,
      "B7oj6ja": _u5dwexa.default,
      "Bbmb7ep": _ukikfbv.default,
      "Beyfa6y": _uhf7ou.default
    }
  },
  sm: {
    "props": {
      "Btl43ni": _u1n69reo.default,
      "B7oj6ja": _u1sx1i.default,
      "Bbmb7ep": _ui28ht.default,
      "Beyfa6y": _uyycwk.default
    }
  },
  md: {
    "props": {
      "Btl43ni": _u1j8h7k.default,
      "B7oj6ja": _u10e4wt.default,
      "Bbmb7ep": _u5rwohh.default,
      "Beyfa6y": _u14kz1ee.default
    }
  },
  lg: {
    "props": {
      "Btl43ni": _uacbfjb.default,
      "B7oj6ja": _u1l86f2g.default,
      "Bbmb7ep": _u1r1mxth.default,
      "Beyfa6y": _ul6wel.default
    }
  },
  xl: {
    "props": {
      "Btl43ni": _u1m59yg.default,
      "B7oj6ja": _ubm82nm.default,
      "Bbmb7ep": _u1qcfi6m.default,
      "Beyfa6y": _u1pc15gj.default
    }
  }
};
