// @universe-design/react/src/progress/stories/homeless.stories.tsx
import React, { useState } from 'react';
import type { Meta } from '@storybook/react';
import { Progress, Space, InputNumber } from '../..';

export default {
  title: 'Components/Progress/临时示例',
  component: Progress,
} satisfies Meta;

export const LineProgress = () => {
  const [percent, setPercent] = useState<number>(0);

  return (
    <Space direction="vertical" size="large" style={{ width: '50%' }}>
      <Progress type="line" percent={percent} />
      <Progress type="line" percent={percent} infoPosition="left" />
      <Progress type="line" percent={percent} status="exception" />
      <Progress type="line" percent={percent} status="success" />
      <Progress type="line" percent={percent} status="normal" />
      <Progress type="line" percent={percent} status="normal" format={p => `自定义：${p} %`} />
      <Progress type="line" percent={percent} showInfo={false} />
      <Progress type="line" percent={percent} status="normal" strokeColor="red" trailColor="yellow" strokeWidth={10} />
      <div>
        <InputNumber min={0} max={100} onChange={e => setPercent(e || 0)} />
        <span>修改 percent</span>
      </div>
    </Space>
  );
};

export const CircleProgress = () => {
  const [percent, setPercent] = useState<number>(0);

  return (
    <>
      <Space size="large">
        <Progress type="circle" percent={percent} showInfo={false} />
        <Progress type="circle" percent={percent} showInfo={true} />
        <Progress type="circle" percent={percent} width={10} />
        <Progress type="circle" percent={percent} infoPosition="right" />
        <Progress type="circle" percent={percent} showInfo={false} />
        <Progress type="circle" percent={percent} status="success" showInfo={false} />
        <Progress type="circle" percent={percent} status="exception" showInfo={false} />
        <Progress type="circle" percent={percent} width={30} strokeWidth={5} strokeColor="red" trailColor="yellow" />
        <Progress type="circle" percent={percent} showInfo={false} format={p => `自定义：${p} %`} />
      </Space>
      <div>
        <InputNumber min={0} max={100} onChange={e => setPercent(e || 0)} />
        <span>修改 percent</span>
      </div>
    </>
  );
};

export const Success = () => {
  const [percent, setPercent] = useState<number>(0);
  return (
    <Space size="large" direction="vertical">
      <Progress type="circle" percent={percent} status={percent >= 100 ? 'success' : 'normal'} />
      <InputNumber min={0} max={100} onChange={e => setPercent(e || 0)} />
    </Space>
  );
};

export const Safe = () => (
  <>
    <Progress percent={-100} />
    <Progress type="circle" percent={-100} />
    <Progress percent={500} />
    <Progress type="circle" percent={500} />
  </>
);

export const LittlePercentage = () => {
  const [percent, setPercent] = useState<number>(0);
  const [strokeWidth, setStrokeWidth] = useState<number>(0);
  const [radius, setRadius] = useState<number>(0);

  return (
    <Space direction="vertical" size="large" style={{ width: '50%' }}>
      <Progress percent={percent} strokeWidth={strokeWidth} />
      <Progress type="circle" percent={percent} showInfo={false} width={radius} strokeWidth={strokeWidth} />
      <div>
        <InputNumber min={0} max={100} onChange={e => setPercent(e || 0)} /> <span>修改 percent</span>
      </div>
      <div>
        <InputNumber min={0} onChange={e => setStrokeWidth(e || 0)} /> <span>修改 strokeWidth</span>
      </div>
      <div>
        <InputNumber min={0} onChange={e => setRadius(e || 0)} /> <span>修改 圆形进度条的直径</span>
      </div>
    </Space>
  );
};
