// @universe-design/react/src/progress/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import CBasic from '../demo/basic';
import CCircle from '../demo/circle';
import CSize from '../demo/size';
import CStatus from '../demo/status';
if (!(CBasic as Story).storyName) {
  (CBasic as Story).storyName = '0. 基础用法';
  (CBasic as Story).parameters = {
    docs: {
      description: {
        story: '简单的用法。\n\n',
      },
    },
  };
}
if (!(CCircle as Story).storyName) {
  (CCircle as Story).storyName = '3. 圆形进度条';
  (CCircle as Story).parameters = {
    docs: {
      description: {
        story: '设置`type`为`circle`，开启圆形进度条模式。\n\n',
      },
    },
  };
}
if (!(CSize as Story).storyName) {
  (CSize as Story).storyName = '2. 自定义宽度';
  (CSize as Story).parameters = {
    docs: {
      description: {
        story: '通过`strokeWidth`自定义进度条。\n\n',
      },
    },
  };
}
if (!(CStatus as Story).storyName) {
  (CStatus as Story).storyName = '1. 进度条状态';
  (CStatus as Story).parameters = {
    docs: {
      description: {
        story: '通过`status`设置进度条的状态。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Progress/官网示例',
  tags: ['autodocs'],
};

export { CBasic, CCircle, CSize, CStatus };
