// @universe-design/styles/cjs/preset-dark.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.G400 = exports.G350 = exports.G300 = exports.G200 = exports.G100 = exports.C900 = exports.C800 = exports.C700 = exports.C600 = exports.C500 = exports.C50 = exports.C400 = exports.C350 = exports.C300 = exports.C200 = exports.C100 = exports.B900 = exports.B800 = exports.B700 = exports.B600 = exports.B500 = exports.B50 = exports.B400 = exports.B350 = exports.B300 = exports.B200 = exports.B100 = exports.Aurora_bg_subordinate = exports.Aurora_bg_reflection = exports.Aurora_bg_main = exports.AI_sendicon = exports.AI_primary_fill_transparent_03 = exports.AI_primary_fill_transparent_02 = exports.AI_primary_fill_transparent_01 = exports.AI_primary_fill_solid_03 = exports.AI_primary_fill_solid_02 = exports.AI_primary_fill_solid_01 = exports.AI_primary_fill_pressed = exports.AI_primary_fill_loading = exports.AI_primary_fill_hover = exports.AI_primary_fill_default = exports.AI_primary_content_pressed = exports.AI_primary_content_loading = exports.AI_primary_content_hover = exports.AI_primary_content_default = exports.AI_loading = exports.AI_icon_in_button = exports.AI_icon_gradient = exports.AI_dynamic_line = exports.AI_cursor = void 0;
exports.N650 = exports.N600 = exports.N500 = exports.N50 = exports.N400 = exports.N350 = exports.N300 = exports.N200 = exports.N1000 = exports.N100 = exports.N00_90 = exports.N00_80 = exports.N00_70 = exports.N00_60 = exports.N00_50 = exports.N00_5 = exports.N00_40 = exports.N00_30 = exports.N00_20 = exports.N00_15 = exports.N00_10 = exports.N00 = exports.L900 = exports.L800 = exports.L700 = exports.L600 = exports.L500 = exports.L50 = exports.L400 = exports.L350 = exports.L300 = exports.L200 = exports.L100 = exports.I900 = exports.I800 = exports.I700 = exports.I600 = exports.I500 = exports.I50 = exports.I400 = exports.I350 = exports.I300 = exports.I200 = exports.I100 = exports.G900 = exports.G800 = exports.G700 = exports.G600 = exports.G500 = exports.G50 = void 0;
exports.S200 = exports.S100 = exports.R900 = exports.R800 = exports.R700 = exports.R600 = exports.R500 = exports.R50 = exports.R400 = exports.R350 = exports.R300 = exports.R200 = exports.R100 = exports.P900 = exports.P800 = exports.P700 = exports.P600 = exports.P500 = exports.P50 = exports.P400 = exports.P350 = exports.P300 = exports.P200 = exports.P100 = exports.O900 = exports.O800 = exports.O700 = exports.O600 = exports.O500 = exports.O50 = exports.O400 = exports.O350 = exports.O300 = exports.O200 = exports.O100 = exports.N950 = exports.N900_90 = exports.N900_80 = exports.N900_70 = exports.N900_60 = exports.N900_50 = exports.N900_5 = exports.N900_40 = exports.N900_30 = exports.N900_20 = exports.N900_15 = exports.N900_10 = exports.N900 = exports.N800 = exports.N700 = void 0;
exports.Y350 = exports.Y300 = exports.Y200 = exports.Y100 = exports.W900 = exports.W800 = exports.W700 = exports.W600 = exports.W500 = exports.W50 = exports.W400 = exports.W350 = exports.W300 = exports.W200 = exports.W100 = exports.Vnext_bg_thirdly_navigation = exports.Vnext_bg_sub_navigation = exports.Vnext_bg_gradient = exports.Vnext_bg_content_header = exports.V900 = exports.V800 = exports.V700 = exports.V600 = exports.V500 = exports.V50 = exports.V400 = exports.V350 = exports.V300 = exports.V200 = exports.V100 = exports.T900 = exports.T800 = exports.T700 = exports.T600 = exports.T500 = exports.T50 = exports.T400 = exports.T350 = exports.T300 = exports.T200 = exports.T100 = exports.S900 = exports.S800 = exports.S700 = exports.S600 = exports.S500 = exports.S50 = exports.S400 = exports.S350 = exports.S300 = void 0;
exports.function_danger_50 = exports.function_danger_400 = exports.function_danger_350 = exports.function_danger_300 = exports.function_danger_200 = exports.function_danger_100 = exports.fill_tag = exports.fill_selected = exports.fill_pressed = exports.fill_loading_mask = exports.fill_img_mask = exports.fill_hover = exports.fill_focus = exports.fill_disabled = exports.fill_active = exports.colorful_yellow = exports.colorful_wathet = exports.colorful_violet = exports.colorful_turquoise = exports.colorful_sunflower = exports.colorful_red = exports.colorful_purple = exports.colorful_orange = exports.colorful_neutral = exports.colorful_lime = exports.colorful_indigo = exports.colorful_green = exports.colorful_carmine = exports.colorful_blue = exports.bg_tips = exports.bg_text_selection = exports.bg_sub_navigation = exports.bg_pricolor = exports.bg_mask = exports.bg_float_push = exports.bg_float_overlay = exports.bg_float_base = exports.bg_float = exports.bg_filler = exports.bg_content_base = exports.bg_body_overlay = exports.bg_body = exports.bg_base = exports.Y900 = exports.Y800 = exports.Y700 = exports.Y600 = exports.Y500 = exports.Y50 = exports.Y400 = void 0;
exports.function_success_350 = exports.function_success_300 = exports.function_success_200 = exports.function_success_100 = exports.function_info_on_info_fill = exports.function_info_fill_transparent_03 = exports.function_info_fill_transparent_02 = exports.function_info_fill_transparent_01 = exports.function_info_fill_solid_03 = exports.function_info_fill_solid_02 = exports.function_info_fill_solid_01 = exports.function_info_fill_pressed = exports.function_info_fill_loading = exports.function_info_fill_hover = exports.function_info_fill_default = exports.function_info_content_pressed = exports.function_info_content_loading = exports.function_info_content_hover = exports.function_info_content_default = exports.function_info_900 = exports.function_info_800 = exports.function_info_700 = exports.function_info_600 = exports.function_info_500 = exports.function_info_50 = exports.function_info_400 = exports.function_info_350 = exports.function_info_300 = exports.function_info_200 = exports.function_info_100 = exports.function_danger_on_danger_fill = exports.function_danger_fill_transparent_03 = exports.function_danger_fill_transparent_02 = exports.function_danger_fill_transparent_01 = exports.function_danger_fill_solid_03 = exports.function_danger_fill_solid_02 = exports.function_danger_fill_solid_01 = exports.function_danger_fill_pressed = exports.function_danger_fill_loading = exports.function_danger_fill_hover = exports.function_danger_fill_default = exports.function_danger_content_pressed = exports.function_danger_content_loading = exports.function_danger_content_hover = exports.function_danger_content_default = exports.function_danger_900 = exports.function_danger_800 = exports.function_danger_700 = exports.function_danger_600 = exports.function_danger_500 = void 0;
exports.gradient_carmine = exports.gradient_blue = exports.function_warning_on_warning_fill = exports.function_warning_fill_transparent_03 = exports.function_warning_fill_transparent_02 = exports.function_warning_fill_transparent_01 = exports.function_warning_fill_solid_03 = exports.function_warning_fill_solid_02 = exports.function_warning_fill_solid_01 = exports.function_warning_fill_pressed = exports.function_warning_fill_loading = exports.function_warning_fill_hover = exports.function_warning_fill_default = exports.function_warning_content_pressed = exports.function_warning_content_loading = exports.function_warning_content_hover = exports.function_warning_content_default = exports.function_warning_900 = exports.function_warning_800 = exports.function_warning_700 = exports.function_warning_600 = exports.function_warning_500 = exports.function_warning_50 = exports.function_warning_400 = exports.function_warning_350 = exports.function_warning_300 = exports.function_warning_200 = exports.function_warning_100 = exports.function_success_on_success_fill = exports.function_success_fill_transparent_03 = exports.function_success_fill_transparent_02 = exports.function_success_fill_transparent_01 = exports.function_success_fill_solid_03 = exports.function_success_fill_solid_02 = exports.function_success_fill_solid_01 = exports.function_success_fill_pressed = exports.function_success_fill_loading = exports.function_success_fill_hover = exports.function_success_fill_default = exports.function_success_content_pressed = exports.function_success_content_loading = exports.function_success_content_hover = exports.function_success_content_default = exports.function_success_900 = exports.function_success_800 = exports.function_success_700 = exports.function_success_600 = exports.function_success_500 = exports.function_success_50 = exports.function_success_400 = void 0;
exports.primary_fill_transparent_02 = exports.primary_fill_transparent_01 = exports.primary_fill_solid_03 = exports.primary_fill_solid_02 = exports.primary_fill_solid_01 = exports.primary_fill_pressed = exports.primary_fill_loading = exports.primary_fill_hover = exports.primary_fill_default = exports.primary_content_pressed = exports.primary_content_loading = exports.primary_content_hover = exports.primary_content_default = exports.line_divider_module = exports.line_divider_default = exports.line_border_component = exports.line_border_card = exports.inverse_primary_content_pressed = exports.inverse_primary_content_hover = exports.inverse_primary_content_default = exports.illustration_yellow = exports.illustration_wathet = exports.illustration_turquoise = exports.illustration_static_white = exports.illustration_static_black = exports.illustration_red = exports.illustration_orange = exports.illustration_neutral_d = exports.illustration_neutral_c = exports.illustration_neutral_b = exports.illustration_neutral_a = exports.illustration_blue_e = exports.illustration_blue_d = exports.illustration_blue_c = exports.illustration_blue_b = exports.illustration_blue_a = exports.icon_n3 = exports.icon_n2 = exports.icon_n1 = exports.icon_disabled = exports.gradient_yellow = exports.gradient_wathet = exports.gradient_violet = exports.gradient_turquoise = exports.gradient_red = exports.gradient_purple = exports.gradient_orange = exports.gradient_lime = exports.gradient_indigo = exports.gradient_green = void 0;
exports.shadow_s4_up = exports.shadow_s4_right = exports.shadow_s4_left = exports.shadow_s4_down_pri = exports.shadow_s4_down = exports.shadow_s3_up = exports.shadow_s3_right = exports.shadow_s3_left = exports.shadow_s3_down_pri = exports.shadow_s3_down = exports.shadow_s2_up = exports.shadow_s2_right = exports.shadow_s2_left = exports.shadow_s2_down_pri = exports.shadow_s2_down = exports.shadow_s1_up = exports.shadow_s1_right = exports.shadow_s1_left = exports.shadow_s1_down_pri = exports.shadow_s1_down = exports.shadow_pri_md_down = exports.shadow_pri_md = exports.shadow_pri_lg_down = exports.shadow_pri_lg = exports.shadow_pri = exports.shadow_md_up = exports.shadow_md_right = exports.shadow_md_left = exports.shadow_md_down = exports.shadow_lg_up = exports.shadow_lg_right = exports.shadow_lg_left = exports.shadow_lg_down = exports.shadow_default_sm = exports.shadow_default_md = exports.shadow_default_lg = exports.shadow_default = exports.primary_pri_900 = exports.primary_pri_800 = exports.primary_pri_700 = exports.primary_pri_600 = exports.primary_pri_500 = exports.primary_pri_50 = exports.primary_pri_400 = exports.primary_pri_350 = exports.primary_pri_300 = exports.primary_pri_200 = exports.primary_pri_100 = exports.primary_on_primary_fill = exports.primary_fill_transparent_03 = void 0;
exports.text_title = exports.text_placeholder = exports.text_link_pressed = exports.text_link_normal = exports.text_link_loading = exports.text_link_hover = exports.text_link_disabled = exports.text_disabled = exports.text_caption = exports.static_white_pressed = exports.static_white_hover = exports.static_white_90 = exports.static_white_80 = exports.static_white_70 = exports.static_white_60 = exports.static_white_50 = exports.static_white_5 = exports.static_white_40 = exports.static_white_30 = exports.static_white_20 = exports.static_white_15 = exports.static_white_10 = exports.static_white = exports.static_black_90 = exports.static_black_80 = exports.static_black_70 = exports.static_black_60 = exports.static_black_50 = exports.static_black_5 = exports.static_black_40 = exports.static_black_30 = exports.static_black_20 = exports.static_black_15 = exports.static_black_10 = exports.static_black = exports.shadow_sm_up = exports.shadow_sm_right = exports.shadow_sm_left = exports.shadow_sm_down = exports.shadow_s5_up = exports.shadow_s5_right = exports.shadow_s5_left = exports.shadow_s5_down_pri = exports.shadow_s5_down = void 0;
// autogen by ../scripts/gen-ts-from-json.ts
exports.AI_cursor = 'rgb(94, 104, 232)';
exports.AI_dynamic_line = 'conic-gradient(from 45deg, rgb(164, 61, 164) 0%, rgb(94, 104, 232) 33.33333333333333%, rgb(94, 104, 232) 66.66666666666666%, rgb(164, 61, 164) 100%)';
exports.AI_icon_gradient = 'linear-gradient(to right, rgb(123, 131, 247) 0%, rgb(140, 85, 236) 100%)';
exports.AI_icon_in_button = 'rgb(123, 131, 247)';
exports.AI_loading = 'linear-gradient(to right, rgb(123, 131, 247) 0%, rgb(164, 61, 164) 100%)';
exports.AI_primary_content_default = 'linear-gradient(to right, rgb(123, 131, 247) 0%, rgb(181, 74, 181) 100%)';
exports.AI_primary_content_hover = 'linear-gradient(to right, rgb(94, 104, 232) 0%, rgb(164, 61, 164) 100%)';
exports.AI_primary_content_loading = 'linear-gradient(to right, rgba(123, 131, 247, 0.6) 0%, rgba(181, 74, 181, 0.6) 100%)';
exports.AI_primary_content_pressed = 'linear-gradient(to right, rgb(148, 153, 247) 0%, rgb(214, 97, 214) 100%)';
exports.AI_primary_fill_default = 'linear-gradient(to right, rgb(94, 104, 232) 0%, rgb(139, 55, 139) 100%)';
exports.AI_primary_fill_hover = 'linear-gradient(to right, rgb(81, 90, 214) 0%, rgb(113, 40, 113) 100%)';
exports.AI_primary_fill_loading = 'linear-gradient(to right, rgba(94, 104, 232, 0.6) 0%, rgba(139, 55, 139, 0.6) 100%)';
exports.AI_primary_fill_pressed = 'linear-gradient(to right, rgb(123, 131, 247) 0%, rgb(164, 61, 164) 100%)';
exports.AI_primary_fill_solid_01 = 'linear-gradient(to right, rgb(30, 32, 74) 0%, rgb(59, 21, 59) 100%)';
exports.AI_primary_fill_solid_02 = 'linear-gradient(to right, rgb(42, 45, 105) 0%, rgb(84, 24, 84) 100%)';
exports.AI_primary_fill_solid_03 = 'linear-gradient(to right, rgb(55, 61, 144) 0%, rgb(113, 40, 113) 100%)';
exports.AI_primary_fill_transparent_01 = 'linear-gradient(to right, rgba(94, 104, 232, 0.15) 0%, rgba(164, 61, 164, 0.15) 100%)';
exports.AI_primary_fill_transparent_02 = 'linear-gradient(to right, rgba(94, 104, 232, 0.25) 0%, rgba(164, 61, 164, 0.25) 100%)';
exports.AI_primary_fill_transparent_03 = 'linear-gradient(to right, rgba(94, 104, 232, 0.35) 0%, rgba(164, 61, 164, 0.35) 100%)';
exports.AI_sendicon = 'linear-gradient(to right, rgb(94, 104, 232) 0%, rgb(140, 85, 236) 100%)';
exports.Aurora_bg_main = 'rgba(16, 153, 204, 0.2)';
exports.Aurora_bg_reflection = 'rgba(165, 117, 250, 0.15)';
exports.Aurora_bg_subordinate = 'rgba(16, 153, 204, 0.2)';
exports.B100 = '#173166';
exports.B200 = '#194294';
exports.B300 = '#2655b6';
exports.B350 = '#275fce';
exports.B400 = '#3370eb';
exports.B50 = '#152340';
exports.B500 = '#4c88ff';
exports.B600 = '#75a4ff';
exports.B700 = '#8fb4ff';
exports.B800 = '#bdd2ff';
exports.B900 = '#e0e9ff';
exports.C100 = '#591c3f';
exports.C200 = '#782b57';
exports.C300 = '#94386c';
exports.C350 = '#ab417d';
exports.C400 = '#c24a8e';
exports.C50 = '#3a182b';
exports.C500 = '#db5ea4';
exports.C600 = '#ed77ba';
exports.C700 = '#fc94cf';
exports.C800 = '#ffc2e5';
exports.C900 = '#ffe0f2';
exports.G100 = '#173b12';
exports.G200 = '#21511a';
exports.G300 = '#296621';
exports.G350 = '#2f7526';
exports.G400 = '#35872a';
exports.G50 = '#0e2b0a';
exports.G500 = '#419e34';
exports.G600 = '#51ba43';
exports.G700 = '#69cc5c';
exports.G800 = '#99e490';
exports.G900 = '#cbf5c6';
exports.I100 = '#2a2d69';
exports.I200 = '#373d90';
exports.I300 = '#474fb8';
exports.I350 = '#515ad6';
exports.I400 = '#5e68e8';
exports.I50 = '#1e204a';
exports.I500 = '#7b83f7';
exports.I600 = '#9499f7';
exports.I700 = '#aaaff8';
exports.I800 = '#cdd0f9';
exports.I900 = '#e7e9fe';
exports.L100 = '#303804';
exports.L200 = '#404c06';
exports.L300 = '#53610a';
exports.L350 = '#5f700a';
exports.L400 = '#6b7f05';
exports.L50 = '#212702';
exports.L500 = '#82990a';
exports.L600 = '#93af04';
exports.L700 = '#a6c313';
exports.L800 = '#c2e12d';
exports.L900 = '#e3f391';
exports.N00 = '#0a0a0a';
exports.N00_10 = 'rgba(10, 10, 10, 0.1)';
exports.N00_15 = 'rgba(10, 10, 10, 0.15)';
exports.N00_20 = 'rgba(10, 10, 10, 0.2)';
exports.N00_30 = 'rgba(10, 10, 10, 0.3)';
exports.N00_40 = 'rgba(10, 10, 10, 0.4)';
exports.N00_5 = 'rgba(10, 10, 10, 0.05)';
exports.N00_50 = 'rgba(10, 10, 10, 0.5)';
exports.N00_60 = 'rgba(10, 10, 10, 0.6)';
exports.N00_70 = 'rgba(10, 10, 10, 0.7)';
exports.N00_80 = 'rgba(10, 10, 10, 0.8)';
exports.N00_90 = 'rgba(10, 10, 10, 0.9)';
exports.N100 = '#292929';
exports.N1000 = '#ffffff';
exports.N200 = '#373737';
exports.N300 = '#434343';
exports.N350 = '#505050';
exports.N400 = '#5f5f5f';
exports.N50 = '#1a1a1a';
exports.N500 = '#757575';
exports.N600 = '#a6a6a6';
exports.N650 = '#cfcfcf';
exports.N700 = '#e0e0e0';
exports.N800 = '#e8e8e8';
exports.N900 = '#ebebeb';
exports.N900_10 = 'rgba(235, 235, 235, 0.1)';
exports.N900_15 = 'rgba(235, 235, 235, 0.15)';
exports.N900_20 = 'rgba(235, 235, 235, 0.2)';
exports.N900_30 = 'rgba(235, 235, 235, 0.3)';
exports.N900_40 = 'rgba(235, 235, 235, 0.4)';
exports.N900_5 = 'rgba(235, 235, 235, 0.05)';
exports.N900_50 = 'rgba(235, 235, 235, 0.5)';
exports.N900_60 = 'rgba(235, 235, 235, 0.6)';
exports.N900_70 = 'rgba(235, 235, 235, 0.7)';
exports.N900_80 = 'rgba(235, 235, 235, 0.8)';
exports.N900_90 = 'rgba(235, 235, 235, 0.9)';
exports.N950 = '#f8f8f8';
exports.O100 = '#4a2b10';
exports.O200 = '#683a12';
exports.O300 = '#8a4a19';
exports.O350 = '#a15317';
exports.O400 = '#b85e1a';
exports.O50 = '#33210b';
exports.O500 = '#db7018';
exports.O600 = '#f3871b';
exports.O700 = '#f89e44';
exports.O800 = '#fec88b';
exports.O900 = '#fee7cd';
exports.P100 = '#3f2073';
exports.P200 = '#5529a3';
exports.P300 = '#6c39c6';
exports.P350 = '#7c4ad4';
exports.P400 = '#8c55ec';
exports.P50 = '#2b194a';
exports.P500 = '#a575fa';
exports.P600 = '#b88ffe';
exports.P700 = '#c5a3ff';
exports.P800 = '#dbc8fd';
exports.P900 = '#efe5ff';
exports.R100 = '#591f1d';
exports.R200 = '#7b2524';
exports.R300 = '#a03331';
exports.R350 = '#b33a37';
exports.R400 = '#d14642';
exports.R50 = '#3d1a19';
exports.R500 = '#f05b56';
exports.R600 = '#f6827e';
exports.R700 = '#f89e9b';
exports.R800 = '#fdc6c4';
exports.R900 = '#fee3e2';
exports.S100 = '#38320c';
exports.S200 = '#574d01';
exports.S300 = '#7a6a01';
exports.S350 = '#9c8702';
exports.S400 = '#c9b218';
exports.S50 = '#29250a';
exports.S500 = '#e5cd17';
exports.S600 = '#f5df36';
exports.S700 = '#f7e663';
exports.S800 = '#faed7a';
exports.S900 = '#fff59e';
exports.T100 = '#173b36';
exports.T200 = '#1d4e47';
exports.T300 = '#25665e';
exports.T350 = '#227369';
exports.T400 = '#198578';
exports.T50 = '#132926';
exports.T500 = '#1fa18f';
exports.T600 = '#1ab7a1';
exports.T700 = '#17cfb5';
exports.T800 = '#65e7d5';
exports.T900 = '#b7f7ef';
exports.V100 = '#541854';
exports.V200 = '#712871';
exports.V300 = '#8b378b';
exports.V350 = '#a43da4';
exports.V400 = '#b54ab5';
exports.V50 = '#3b153b';
exports.V500 = '#d661d6';
exports.V600 = '#e17fe1';
exports.V700 = '#e99be9';
exports.V800 = '#f4c3f4';
exports.V900 = '#fcdffc';
exports.Vnext_bg_content_header = 'rgb(34, 34, 34)';
exports.Vnext_bg_gradient = 'linear-gradient(180deg, rgba(26, 26, 26, 0.05) 0%, rgba(10, 10, 10, 0.05) 100%)';
exports.Vnext_bg_sub_navigation = 'rgba(30, 30, 30, 0.5)';
exports.Vnext_bg_thirdly_navigation = 'rgba(26, 26, 26, 0.7)';
exports.W100 = '#103647';
exports.W200 = '#164359';
exports.W300 = '#135a78';
exports.W350 = '#146c91';
exports.W400 = '#1a7fab';
exports.W50 = '#152830';
exports.W500 = '#1099cc';
exports.W600 = '#25b2e5';
exports.W700 = '#51c3f0';
exports.W800 = '#89dffe';
exports.W900 = '#c7f0ff';
exports.Y100 = '#473409';
exports.Y200 = '#63470f';
exports.Y300 = '#8a6419';
exports.Y350 = '#ad7d15';
exports.Y400 = '#d49b0b';
exports.Y50 = '#30250a';
exports.Y500 = '#f0b622';
exports.Y600 = '#fbcb46';
exports.Y700 = '#fcd456';
exports.Y800 = '#ffde75';
exports.Y900 = '#ffeaa3';
exports.bg_base = 'rgb(10, 10, 10)';
exports.bg_body = 'rgb(26, 26, 26)';
exports.bg_body_overlay = 'rgb(41, 41, 41)';
exports.bg_content_base = 'rgb(18, 18, 18)';
exports.bg_filler = 'rgb(55, 55, 55)';
exports.bg_float = 'rgb(41, 41, 41)';
exports.bg_float_base = 'rgb(26, 26, 26)';
exports.bg_float_overlay = 'rgb(55, 55, 55)';
exports.bg_float_push = 'rgba(41, 41, 41, 0.8)';
exports.bg_mask = 'rgba(0, 0, 0, 0.6)';
exports.bg_pricolor = 'rgb(51, 112, 235)';
exports.bg_sub_navigation = 'rgb(38, 38, 38)';
exports.bg_text_selection = 'rgba(117, 164, 255, 0.3)';
exports.bg_tips = 'rgb(80, 80, 80)';
exports.colorful_blue = 'rgb(76, 136, 255)';
exports.colorful_carmine = 'rgb(194, 74, 142)';
exports.colorful_green = 'rgb(65, 158, 52)';
exports.colorful_indigo = 'rgb(94, 104, 232)';
exports.colorful_lime = 'rgb(147, 175, 4)';
exports.colorful_neutral = 'rgb(117, 117, 117)';
exports.colorful_orange = 'rgb(243, 135, 27)';
exports.colorful_purple = 'rgb(140, 85, 236)';
exports.colorful_red = 'rgb(240, 91, 86)';
exports.colorful_sunflower = 'rgb(229, 205, 23)';
exports.colorful_turquoise = 'rgb(31, 161, 143)';
exports.colorful_violet = 'rgb(181, 74, 181)';
exports.colorful_wathet = 'rgb(16, 153, 204)';
exports.colorful_yellow = 'rgb(240, 182, 34)';
exports.fill_active = 'rgba(76, 136, 255, 0.2)';
exports.fill_disabled = 'rgb(95, 95, 95)';
exports.fill_focus = 'rgba(235, 235, 235, 0.12)';
exports.fill_hover = 'rgba(235, 235, 235, 0.08)';
exports.fill_img_mask = 'rgba(0, 0, 0, 0.12)';
exports.fill_loading_mask = 'rgba(10, 10, 10, 0.6)';
exports.fill_pressed = 'rgba(235, 235, 235, 0.12)';
exports.fill_selected = 'rgba(76, 136, 255, 0.15)';
exports.fill_tag = 'rgba(235, 235, 235, 0.1)';
exports.function_danger_100 = 'rgb(89, 31, 29)';
exports.function_danger_200 = 'rgb(123, 37, 36)';
exports.function_danger_300 = 'rgb(160, 51, 49)';
exports.function_danger_350 = 'rgb(179, 58, 55)';
exports.function_danger_400 = 'rgb(209, 70, 66)';
exports.function_danger_50 = 'rgb(61, 26, 25)';
exports.function_danger_500 = 'rgb(240, 91, 86)';
exports.function_danger_600 = 'rgb(246, 130, 126)';
exports.function_danger_700 = 'rgb(248, 158, 155)';
exports.function_danger_800 = 'rgb(253, 198, 196)';
exports.function_danger_900 = 'rgb(254, 227, 226)';
exports.function_danger_content_default = 'rgb(240, 91, 86)';
exports.function_danger_content_hover = 'rgb(209, 70, 66)';
exports.function_danger_content_loading = 'rgb(160, 51, 49)';
exports.function_danger_content_pressed = 'rgb(246, 130, 126)';
exports.function_danger_fill_default = 'rgb(209, 70, 66)';
exports.function_danger_fill_hover = 'rgb(179, 58, 55)';
exports.function_danger_fill_loading = 'rgb(160, 51, 49)';
exports.function_danger_fill_pressed = 'rgb(240, 91, 86)';
exports.function_danger_fill_solid_01 = 'rgb(61, 26, 25)';
exports.function_danger_fill_solid_02 = 'rgb(89, 31, 29)';
exports.function_danger_fill_solid_03 = 'rgb(123, 37, 36)';
exports.function_danger_fill_transparent_01 = 'rgba(240, 91, 86, 0.15)';
exports.function_danger_fill_transparent_02 = 'rgba(240, 91, 86, 0.2)';
exports.function_danger_fill_transparent_03 = 'rgba(240, 91, 86, 0.3)';
exports.function_danger_on_danger_fill = 'rgb(255, 255, 255)';
exports.function_info_100 = 'rgb(23, 49, 102)';
exports.function_info_200 = 'rgb(25, 66, 148)';
exports.function_info_300 = 'rgb(38, 85, 182)';
exports.function_info_350 = 'rgb(39, 95, 206)';
exports.function_info_400 = 'rgb(51, 112, 235)';
exports.function_info_50 = 'rgb(21, 35, 64)';
exports.function_info_500 = 'rgb(76, 136, 255)';
exports.function_info_600 = 'rgb(117, 164, 255)';
exports.function_info_700 = 'rgb(143, 180, 255)';
exports.function_info_800 = 'rgb(189, 210, 255)';
exports.function_info_900 = 'rgb(224, 233, 255)';
exports.function_info_content_default = 'rgb(76, 136, 255)';
exports.function_info_content_hover = 'rgb(51, 112, 235)';
exports.function_info_content_loading = 'rgb(38, 85, 182)';
exports.function_info_content_pressed = 'rgb(117, 164, 255)';
exports.function_info_fill_default = 'rgb(51, 112, 235)';
exports.function_info_fill_hover = 'rgb(39, 95, 206)';
exports.function_info_fill_loading = 'rgb(38, 85, 182)';
exports.function_info_fill_pressed = 'rgb(76, 136, 255)';
exports.function_info_fill_solid_01 = 'rgb(21, 35, 64)';
exports.function_info_fill_solid_02 = 'rgb(23, 49, 102)';
exports.function_info_fill_solid_03 = 'rgb(25, 66, 148)';
exports.function_info_fill_transparent_01 = 'rgba(76, 136, 255, 0.15)';
exports.function_info_fill_transparent_02 = 'rgba(76, 136, 255, 0.2)';
exports.function_info_fill_transparent_03 = 'rgba(76, 136, 255, 0.3)';
exports.function_info_on_info_fill = 'rgb(255, 255, 255)';
exports.function_success_100 = 'rgb(23, 59, 18)';
exports.function_success_200 = 'rgb(33, 81, 26)';
exports.function_success_300 = 'rgb(41, 102, 33)';
exports.function_success_350 = 'rgb(47, 117, 38)';
exports.function_success_400 = 'rgb(53, 135, 42)';
exports.function_success_50 = 'rgb(14, 43, 10)';
exports.function_success_500 = 'rgb(65, 158, 52)';
exports.function_success_600 = 'rgb(81, 186, 67)';
exports.function_success_700 = 'rgb(105, 204, 92)';
exports.function_success_800 = 'rgb(153, 228, 144)';
exports.function_success_900 = 'rgb(203, 245, 198)';
exports.function_success_content_default = 'rgb(81, 186, 67)';
exports.function_success_content_hover = 'rgb(65, 158, 52)';
exports.function_success_content_loading = 'rgb(41, 102, 33)';
exports.function_success_content_pressed = 'rgb(105, 204, 92)';
exports.function_success_fill_default = 'rgb(65, 158, 52)';
exports.function_success_fill_hover = 'rgb(53, 135, 42)';
exports.function_success_fill_loading = 'rgb(41, 102, 33)';
exports.function_success_fill_pressed = 'rgb(81, 186, 67)';
exports.function_success_fill_solid_01 = 'rgb(14, 43, 10)';
exports.function_success_fill_solid_02 = 'rgb(23, 59, 18)';
exports.function_success_fill_solid_03 = 'rgb(33, 81, 26)';
exports.function_success_fill_transparent_01 = 'rgba(65, 158, 52, 0.15)';
exports.function_success_fill_transparent_02 = 'rgba(65, 158, 52, 0.2)';
exports.function_success_fill_transparent_03 = 'rgba(65, 158, 52, 0.3)';
exports.function_success_on_success_fill = 'rgb(255, 255, 255)';
exports.function_warning_100 = 'rgb(74, 43, 16)';
exports.function_warning_200 = 'rgb(104, 58, 18)';
exports.function_warning_300 = 'rgb(138, 74, 25)';
exports.function_warning_350 = 'rgb(161, 83, 23)';
exports.function_warning_400 = 'rgb(184, 94, 26)';
exports.function_warning_50 = 'rgb(51, 33, 11)';
exports.function_warning_500 = 'rgb(219, 112, 24)';
exports.function_warning_600 = 'rgb(243, 135, 27)';
exports.function_warning_700 = 'rgb(248, 158, 68)';
exports.function_warning_800 = 'rgb(254, 200, 139)';
exports.function_warning_900 = 'rgb(254, 231, 205)';
exports.function_warning_content_default = 'rgb(243, 135, 27)';
exports.function_warning_content_hover = 'rgb(219, 112, 24)';
exports.function_warning_content_loading = 'rgb(138, 74, 25)';
exports.function_warning_content_pressed = 'rgb(248, 158, 68)';
exports.function_warning_fill_default = 'rgb(243, 135, 27)';
exports.function_warning_fill_hover = 'rgb(219, 112, 24)';
exports.function_warning_fill_loading = 'rgb(138, 74, 25)';
exports.function_warning_fill_pressed = 'rgb(248, 158, 68)';
exports.function_warning_fill_solid_01 = 'rgb(51, 33, 11)';
exports.function_warning_fill_solid_02 = 'rgb(74, 43, 16)';
exports.function_warning_fill_solid_03 = 'rgb(104, 58, 18)';
exports.function_warning_fill_transparent_01 = 'rgba(243, 135, 27, 0.15)';
exports.function_warning_fill_transparent_02 = 'rgba(243, 135, 27, 0.2)';
exports.function_warning_fill_transparent_03 = 'rgba(243, 135, 27, 0.3)';
exports.function_warning_on_warning_fill = 'rgb(255, 255, 255)';
exports.gradient_blue = 'linear-gradient(90deg, rgb(51, 112, 235) 0%, rgb(39, 95, 206) 100%)';
exports.gradient_carmine = 'linear-gradient(90deg, rgb(219, 94, 164) 0%, rgb(194, 74, 142) 100%)';
exports.gradient_green = 'linear-gradient(90deg, rgb(65, 158, 52) 0%, rgb(53, 135, 42) 100%)';
exports.gradient_indigo = 'linear-gradient(90deg, rgb(94, 104, 232) 0%, rgb(81, 90, 214) 100%)';
exports.gradient_lime = 'linear-gradient(90deg, rgb(147, 175, 4) 0%, rgb(130, 153, 10) 100%)';
exports.gradient_orange = 'linear-gradient(90deg, rgb(243, 135, 27) 0%, rgb(219, 112, 24) 100%)';
exports.gradient_purple = 'linear-gradient(90deg, rgb(140, 85, 236) 0%, rgb(124, 74, 212) 100%)';
exports.gradient_red = 'linear-gradient(90deg, rgb(240, 91, 86) 0%, rgb(209, 70, 66) 100%)';
exports.gradient_turquoise = 'linear-gradient(90deg, rgb(26, 183, 161) 0%, rgb(31, 161, 143) 100%)';
exports.gradient_violet = 'linear-gradient(90deg, rgb(214, 97, 214) 0%, rgb(181, 74, 181) 100%)';
exports.gradient_wathet = 'linear-gradient(90deg, rgb(16, 153, 204) 0%, rgb(26, 127, 171) 100%)';
exports.gradient_yellow = 'linear-gradient(90deg, rgb(240, 182, 34) 0%, rgb(212, 155, 11) 100%)';
exports.icon_disabled = 'rgb(95, 95, 95)';
exports.icon_n1 = 'rgb(232, 232, 232)';
exports.icon_n2 = 'rgb(166, 166, 166)';
exports.icon_n3 = 'rgb(117, 117, 117)';
exports.illustration_blue_a = 'rgb(224, 224, 224)';
exports.illustration_blue_b = 'rgba(255, 255, 255, 0.4)';
exports.illustration_blue_c = 'rgb(95, 95, 95)';
exports.illustration_blue_d = 'rgb(38, 85, 182)';
exports.illustration_blue_e = 'rgb(76, 136, 255)';
exports.illustration_neutral_a = 'rgb(235, 235, 235)';
exports.illustration_neutral_b = 'rgb(55, 55, 55)';
exports.illustration_neutral_c = 'rgba(255, 255, 255, 0.3)';
exports.illustration_neutral_d = 'rgba(255, 255, 255, 0.4)';
exports.illustration_orange = 'rgb(243, 135, 27)';
exports.illustration_red = 'rgb(240, 91, 86)';
exports.illustration_static_black = 'rgb(0, 0, 0)';
exports.illustration_static_white = 'rgb(255, 255, 255)';
exports.illustration_turquoise = 'rgb(23, 207, 181)';
exports.illustration_wathet = 'rgb(81, 195, 240)';
exports.illustration_yellow = 'rgb(240, 182, 34)';
exports.inverse_primary_content_default = 'rgb(117, 164, 255)';
exports.inverse_primary_content_hover = 'rgb(143, 180, 255)';
exports.inverse_primary_content_pressed = 'rgb(76, 136, 255)';
exports.line_border_card = 'rgba(235, 235, 235, 0.15)';
exports.line_border_component = 'rgb(80, 80, 80)';
exports.line_divider_default = 'rgba(207, 207, 207, 0.15)';
exports.line_divider_module = 'rgb(0, 0, 0)';
exports.primary_content_default = 'rgb(76, 136, 255)';
exports.primary_content_hover = 'rgb(51, 112, 235)';
exports.primary_content_loading = 'rgb(38, 85, 182)';
exports.primary_content_pressed = 'rgb(117, 164, 255)';
exports.primary_fill_default = 'rgb(51, 112, 235)';
exports.primary_fill_hover = 'rgb(39, 95, 206)';
exports.primary_fill_loading = 'rgb(38, 85, 182)';
exports.primary_fill_pressed = 'rgb(76, 136, 255)';
exports.primary_fill_solid_01 = 'rgb(21, 35, 64)';
exports.primary_fill_solid_02 = 'rgb(23, 49, 102)';
exports.primary_fill_solid_03 = 'rgb(25, 66, 148)';
exports.primary_fill_transparent_01 = 'rgba(76, 136, 255, 0.15)';
exports.primary_fill_transparent_02 = 'rgba(76, 136, 255, 0.2)';
exports.primary_fill_transparent_03 = 'rgba(76, 136, 255, 0.3)';
exports.primary_on_primary_fill = 'rgb(255, 255, 255)';
exports.primary_pri_100 = 'rgb(23, 49, 102)';
exports.primary_pri_200 = 'rgb(25, 66, 148)';
exports.primary_pri_300 = 'rgb(38, 85, 182)';
exports.primary_pri_350 = 'rgb(39, 95, 206)';
exports.primary_pri_400 = 'rgb(51, 112, 235)';
exports.primary_pri_50 = 'rgb(21, 35, 64)';
exports.primary_pri_500 = 'rgb(76, 136, 255)';
exports.primary_pri_600 = 'rgb(117, 164, 255)';
exports.primary_pri_700 = 'rgb(143, 180, 255)';
exports.primary_pri_800 = 'rgb(189, 210, 255)';
exports.primary_pri_900 = 'rgb(224, 233, 255)';
exports.shadow_default = 'rgb(0, 0, 0)';
exports.shadow_default_lg = 'rgba(0, 0, 0, 0.24)';
exports.shadow_default_md = 'rgba(0, 0, 0, 0.28)';
exports.shadow_default_sm = 'rgba(0, 0, 0, 0.32)';
exports.shadow_lg_down = '0px 12px 48px 0px rgba(0, 0, 0, 0.24)';
exports.shadow_lg_left = '-12px 0px 48px 0px rgba(0, 0, 0, 0.24)';
exports.shadow_lg_right = '12px 0px 48px 0px rgba(0, 0, 0, 0.24)';
exports.shadow_lg_up = '0px -12px 48px 0px rgba(0, 0, 0, 0.24)';
exports.shadow_md_down = '0px 8px 16px 0px rgba(0, 0, 0, 0.28)';
exports.shadow_md_left = '-8px 0px 16px 0px rgba(0, 0, 0, 0.28)';
exports.shadow_md_right = '8px 0px 16px 0px rgba(0, 0, 0, 0.28)';
exports.shadow_md_up = '0px -8px 16px 0px rgba(0, 0, 0, 0.28)';
exports.shadow_pri = 'rgb(36, 91, 219)';
exports.shadow_pri_lg = 'rgba(25, 66, 148, 0.48)';
exports.shadow_pri_lg_down = '0px 12px 48px 0px rgba(25, 66, 148, 0.48)';
exports.shadow_pri_md = 'rgba(25, 66, 148, 0.48)';
exports.shadow_pri_md_down = '0px 8px 16px 0px rgba(25, 66, 148, 0.48)';
exports.shadow_s1_down = '0px 1px 2px -2px rgba(0, 0, 0, 0.02), 0px 2px 4px 0px rgba(0, 0, 0, 0.02), 0px 2px 8px 2px rgba(0, 0, 0, 0.02)';
exports.shadow_s1_down_pri = '0px 1px 2px -2px rgba(36, 91, 219, 0.02), 0px 2px 4px 0px rgba(36, 91, 219, 0.02), 0px 2px 8px 2px rgba(36, 91, 219, 0.02)';
exports.shadow_s1_left = '-1px 0px 2px -2px rgba(0, 0, 0, 0.02), -2px 0px 4px 0px rgba(0, 0, 0, 0.02), -2px 0px 8px 2px rgba(0, 0, 0, 0.02)';
exports.shadow_s1_right = '1px 0px 2px -2px rgba(0, 0, 0, 0.02), 2px 0px 4px 0px rgba(0, 0, 0, 0.02), 2px 0px 8px 2px rgba(0, 0, 0, 0.02)';
exports.shadow_s1_up = '0px -1px 2px -2px rgba(0, 0, 0, 0.02), 0px -2px 4px 0px rgba(0, 0, 0, 0.02), 0px -2px 8px 2px rgba(0, 0, 0, 0.02)';
exports.shadow_s2_down = '0px 4px 16px 4px rgba(0, 0, 0, 0.03), 0px 4px 8px 0px rgba(0, 0, 0, 0.02), 0px 2px 4px -4px rgba(0, 0, 0, 0.02)';
exports.shadow_s2_down_pri = '0px 4px 16px 4px rgba(36, 91, 219, 0.03), 0px 4px 8px 0px rgba(36, 91, 219, 0.02), 0px 2px 4px -4px rgba(36, 91, 219, 0.02)';
exports.shadow_s2_left = '-4px 0px 16px 4px rgba(0, 0, 0, 0.03), -4px 0px 8px 0px rgba(0, 0, 0, 0.02), -2px 0px 4px -4px rgba(0, 0, 0, 0.02)';
exports.shadow_s2_right = '4px 0px 16px 4px rgba(0, 0, 0, 0.03), 4px 0px 8px 0px rgba(0, 0, 0, 0.02), 2px 0px 4px -4px rgba(0, 0, 0, 0.02)';
exports.shadow_s2_up = '0px -4px 16px 4px rgba(0, 0, 0, 0.03), 0px -4px 8px 0px rgba(0, 0, 0, 0.02), 0px -2px 4px -4px rgba(0, 0, 0, 0.02)';
exports.shadow_s3_down = '0px 6px 18px 6px rgba(0, 0, 0, 0.03), 0px 3px 6px -6px rgba(0, 0, 0, 0.05), 0px 4px 8px 0px rgba(0, 0, 0, 0.03)';
exports.shadow_s3_down_pri = '0px 6px 18px 6px rgba(36, 91, 219, 0.03), 0px 3px 6px -6px rgba(36, 91, 219, 0.05), 0px 4px 8px 0px rgba(36, 91, 219, 0.03)';
exports.shadow_s3_left = '-6px 0px 18px 6px rgba(0, 0, 0, 0.03), -3px 0px 6px -6px rgba(0, 0, 0, 0.05), -4px 0px 8px 0px rgba(0, 0, 0, 0.03)';
exports.shadow_s3_right = '6px 0px 18px 6px rgba(0, 0, 0, 0.03), 3px 0px 6px -6px rgba(0, 0, 0, 0.05), 4px 0px 8px 0px rgba(0, 0, 0, 0.03)';
exports.shadow_s3_up = '0px -6px 18px 6px rgba(0, 0, 0, 0.03), 0px -3px 6px -6px rgba(0, 0, 0, 0.05), 0px -4px 8px 0px rgba(0, 0, 0, 0.03)';
exports.shadow_s4_down = '0px 8px 24px 8px rgba(0, 0, 0, 0.04), 0px 6px 12px 0px rgba(0, 0, 0, 0.04), 0px 4px 8px -8px rgba(0, 0, 0, 0.06)';
exports.shadow_s4_down_pri = '0px 8px 24px 8px rgba(36, 91, 219, 0.04), 0px 6px 12px 0px rgba(36, 91, 219, 0.04), 0px 4px 8px -8px rgba(36, 91, 219, 0.06)';
exports.shadow_s4_left = '-8px 0px 24px 8px rgba(0, 0, 0, 0.04), -6px 0px 12px 0px rgba(0, 0, 0, 0.04), -4px 0px 8px -8px rgba(0, 0, 0, 0.06)';
exports.shadow_s4_right = '8px 0px 24px 8px rgba(0, 0, 0, 0.04), 6px 0px 12px 0px rgba(0, 0, 0, 0.04), 4px 0px 8px -8px rgba(0, 0, 0, 0.06)';
exports.shadow_s4_up = '0px -8px 24px 8px rgba(0, 0, 0, 0.04), 0px -6px 12px 0px rgba(0, 0, 0, 0.04), 0px -4px 8px -8px rgba(0, 0, 0, 0.06)';
exports.shadow_s5_down = '0px 10px 36px 10px rgba(0, 0, 0, 0.04), 0px 8px 24px 0px rgba(0, 0, 0, 0.04), 0px 6px 12px -10px rgba(0, 0, 0, 0.06)';
exports.shadow_s5_down_pri = '0px 10px 36px 10px rgba(36, 91, 219, 0.04), 0px 8px 24px 0px rgba(36, 91, 219, 0.04), 0px 6px 12px -10px rgba(36, 91, 219, 0.06)';
exports.shadow_s5_left = '-10px 0px 36px 10px rgba(0, 0, 0, 0.04), -8px 0px 24px 0px rgba(0, 0, 0, 0.04), -6px 0px 12px -10px rgba(0, 0, 0, 0.06)';
exports.shadow_s5_right = '10px 0px 36px 10px rgba(0, 0, 0, 0.04), 8px 0px 24px 0px rgba(0, 0, 0, 0.04), 6px 0px 12px -10px rgba(0, 0, 0, 0.06)';
exports.shadow_s5_up = '0px -10px 36px 10px rgba(0, 0, 0, 0.04), 0px -8px 24px 0px rgba(0, 0, 0, 0.04), 0px -6px 12px -10px rgba(0, 0, 0, 0.06)';
exports.shadow_sm_down = '0px 4px 8px 0px rgba(0, 0, 0, 0.32)';
exports.shadow_sm_left = '-4px 0px 8px 0px rgba(0, 0, 0, 0.32)';
exports.shadow_sm_right = '4px 0px 8px 0px rgba(0, 0, 0, 0.32)';
exports.shadow_sm_up = '0px -4px 8px 0px rgba(0, 0, 0, 0.32)';
exports.static_black = 'rgb(0, 0, 0)';
exports.static_black_10 = 'rgba(0, 0, 0, 0.1)';
exports.static_black_15 = 'rgba(0, 0, 0, 0.15)';
exports.static_black_20 = 'rgba(0, 0, 0, 0.2)';
exports.static_black_30 = 'rgba(0, 0, 0, 0.3)';
exports.static_black_40 = 'rgba(0, 0, 0, 0.4)';
exports.static_black_5 = 'rgba(0, 0, 0, 0.05)';
exports.static_black_50 = 'rgba(0, 0, 0, 0.5)';
exports.static_black_60 = 'rgba(0, 0, 0, 0.6)';
exports.static_black_70 = 'rgba(0, 0, 0, 0.7)';
exports.static_black_80 = 'rgba(0, 0, 0, 0.8)';
exports.static_black_90 = 'rgba(0, 0, 0, 0.9)';
exports.static_white = 'rgb(255, 255, 255)';
exports.static_white_10 = 'rgba(255, 255, 255, 0.1)';
exports.static_white_15 = 'rgba(255, 255, 255, 0.15)';
exports.static_white_20 = 'rgba(255, 255, 255, 0.2)';
exports.static_white_30 = 'rgba(255, 255, 255, 0.3)';
exports.static_white_40 = 'rgba(255, 255, 255, 0.4)';
exports.static_white_5 = 'rgba(255, 255, 255, 0.05)';
exports.static_white_50 = 'rgba(255, 255, 255, 0.5)';
exports.static_white_60 = 'rgba(255, 255, 255, 0.6)';
exports.static_white_70 = 'rgba(255, 255, 255, 0.7)';
exports.static_white_80 = 'rgba(255, 255, 255, 0.8)';
exports.static_white_90 = 'rgba(255, 255, 255, 0.9)';
exports.static_white_hover = 'rgba(255, 255, 255, 0.1)';
exports.static_white_pressed = 'rgba(255, 255, 255, 0.2)';
exports.text_caption = 'rgb(166, 166, 166)';
exports.text_disabled = 'rgb(95, 95, 95)';
exports.text_link_disabled = 'rgb(95, 95, 95)';
exports.text_link_hover = 'rgb(51, 112, 235)';
exports.text_link_loading = 'rgb(38, 85, 182)';
exports.text_link_normal = 'rgb(76, 136, 255)';
exports.text_link_pressed = 'rgb(117, 164, 255)';
exports.text_placeholder = 'rgb(117, 117, 117)';
exports.text_title = 'rgb(235, 235, 235)';
