// @universe-design/styles/cjs/preset-dark.atom-raw.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.L350_raw = exports.L300_raw = exports.L200_raw = exports.L100_raw = exports.I900_raw = exports.I800_raw = exports.I700_raw = exports.I600_raw = exports.I500_raw = exports.I50_raw = exports.I400_raw = exports.I350_raw = exports.I300_raw = exports.I200_raw = exports.I100_raw = exports.G900_raw = exports.G800_raw = exports.G700_raw = exports.G600_raw = exports.G500_raw = exports.G50_raw = exports.G400_raw = exports.G350_raw = exports.G300_raw = exports.G200_raw = exports.G100_raw = exports.C900_raw = exports.C800_raw = exports.C700_raw = exports.C600_raw = exports.C500_raw = exports.C50_raw = exports.C400_raw = exports.C350_raw = exports.C300_raw = exports.C200_raw = exports.C100_raw = exports.B900_raw = exports.B800_raw = exports.B700_raw = exports.B600_raw = exports.B500_raw = exports.B50_raw = exports.B400_raw = exports.B350_raw = exports.B300_raw = exports.B200_raw = exports.B100_raw = exports.AI_icon_in_button_raw = exports.AI_cursor_raw = void 0;
exports.R50_raw = exports.R400_raw = exports.R350_raw = exports.R300_raw = exports.R200_raw = exports.R100_raw = exports.P900_raw = exports.P800_raw = exports.P700_raw = exports.P600_raw = exports.P500_raw = exports.P50_raw = exports.P400_raw = exports.P350_raw = exports.P300_raw = exports.P200_raw = exports.P100_raw = exports.O900_raw = exports.O800_raw = exports.O700_raw = exports.O600_raw = exports.O500_raw = exports.O50_raw = exports.O400_raw = exports.O350_raw = exports.O300_raw = exports.O200_raw = exports.O100_raw = exports.N950_raw = exports.N900_raw = exports.N800_raw = exports.N700_raw = exports.N650_raw = exports.N600_raw = exports.N500_raw = exports.N50_raw = exports.N400_raw = exports.N350_raw = exports.N300_raw = exports.N200_raw = exports.N1000_raw = exports.N100_raw = exports.N00_raw = exports.L900_raw = exports.L800_raw = exports.L700_raw = exports.L600_raw = exports.L500_raw = exports.L50_raw = exports.L400_raw = void 0;
exports.W700_raw = exports.W600_raw = exports.W500_raw = exports.W50_raw = exports.W400_raw = exports.W350_raw = exports.W300_raw = exports.W200_raw = exports.W100_raw = exports.Vnext_bg_thirdly_navigation_raw = exports.Vnext_bg_sub_navigation_raw = exports.Vnext_bg_content_header_raw = exports.V900_raw = exports.V800_raw = exports.V700_raw = exports.V600_raw = exports.V500_raw = exports.V50_raw = exports.V400_raw = exports.V350_raw = exports.V300_raw = exports.V200_raw = exports.V100_raw = exports.T900_raw = exports.T800_raw = exports.T700_raw = exports.T600_raw = exports.T500_raw = exports.T50_raw = exports.T400_raw = exports.T350_raw = exports.T300_raw = exports.T200_raw = exports.T100_raw = exports.S900_raw = exports.S800_raw = exports.S700_raw = exports.S600_raw = exports.S500_raw = exports.S50_raw = exports.S400_raw = exports.S350_raw = exports.S300_raw = exports.S200_raw = exports.S100_raw = exports.R900_raw = exports.R800_raw = exports.R700_raw = exports.R600_raw = exports.R500_raw = void 0;
exports.function_danger_800_raw = exports.function_danger_700_raw = exports.function_danger_600_raw = exports.function_danger_500_raw = exports.function_danger_50_raw = exports.function_danger_400_raw = exports.function_danger_350_raw = exports.function_danger_300_raw = exports.function_danger_200_raw = exports.function_danger_100_raw = exports.fill_disabled_raw = exports.colorful_yellow_raw = exports.colorful_wathet_raw = exports.colorful_violet_raw = exports.colorful_turquoise_raw = exports.colorful_sunflower_raw = exports.colorful_red_raw = exports.colorful_purple_raw = exports.colorful_orange_raw = exports.colorful_neutral_raw = exports.colorful_lime_raw = exports.colorful_indigo_raw = exports.colorful_green_raw = exports.colorful_carmine_raw = exports.colorful_blue_raw = exports.bg_tips_raw = exports.bg_sub_navigation_raw = exports.bg_pricolor_raw = exports.bg_mask_raw = exports.bg_float_raw = exports.bg_float_overlay_raw = exports.bg_float_base_raw = exports.bg_filler_raw = exports.bg_content_base_raw = exports.bg_body_raw = exports.bg_body_overlay_raw = exports.bg_base_raw = exports.Y900_raw = exports.Y800_raw = exports.Y700_raw = exports.Y600_raw = exports.Y500_raw = exports.Y50_raw = exports.Y400_raw = exports.Y350_raw = exports.Y300_raw = exports.Y200_raw = exports.Y100_raw = exports.W900_raw = exports.W800_raw = void 0;
exports.function_success_content_loading_raw = exports.function_success_content_hover_raw = exports.function_success_content_default_raw = exports.function_success_900_raw = exports.function_success_800_raw = exports.function_success_700_raw = exports.function_success_600_raw = exports.function_success_500_raw = exports.function_success_50_raw = exports.function_success_400_raw = exports.function_success_350_raw = exports.function_success_300_raw = exports.function_success_200_raw = exports.function_success_100_raw = exports.function_info_on_info_fill_raw = exports.function_info_fill_solid_03_raw = exports.function_info_fill_solid_02_raw = exports.function_info_fill_solid_01_raw = exports.function_info_fill_pressed_raw = exports.function_info_fill_loading_raw = exports.function_info_fill_hover_raw = exports.function_info_fill_default_raw = exports.function_info_content_pressed_raw = exports.function_info_content_loading_raw = exports.function_info_content_hover_raw = exports.function_info_content_default_raw = exports.function_info_900_raw = exports.function_info_800_raw = exports.function_info_700_raw = exports.function_info_600_raw = exports.function_info_500_raw = exports.function_info_50_raw = exports.function_info_400_raw = exports.function_info_350_raw = exports.function_info_300_raw = exports.function_info_200_raw = exports.function_info_100_raw = exports.function_danger_on_danger_fill_raw = exports.function_danger_fill_solid_03_raw = exports.function_danger_fill_solid_02_raw = exports.function_danger_fill_solid_01_raw = exports.function_danger_fill_pressed_raw = exports.function_danger_fill_loading_raw = exports.function_danger_fill_hover_raw = exports.function_danger_fill_default_raw = exports.function_danger_content_pressed_raw = exports.function_danger_content_loading_raw = exports.function_danger_content_hover_raw = exports.function_danger_content_default_raw = exports.function_danger_900_raw = void 0;
exports.illustration_wathet_raw = exports.illustration_turquoise_raw = exports.illustration_static_white_raw = exports.illustration_static_black_raw = exports.illustration_red_raw = exports.illustration_orange_raw = exports.illustration_neutral_d_raw = exports.illustration_neutral_b_raw = exports.illustration_neutral_a_raw = exports.illustration_blue_e_raw = exports.illustration_blue_d_raw = exports.illustration_blue_c_raw = exports.illustration_blue_b_raw = exports.illustration_blue_a_raw = exports.icon_n3_raw = exports.icon_n2_raw = exports.icon_n1_raw = exports.icon_disabled_raw = exports.function_warning_on_warning_fill_raw = exports.function_warning_fill_solid_03_raw = exports.function_warning_fill_solid_02_raw = exports.function_warning_fill_solid_01_raw = exports.function_warning_fill_pressed_raw = exports.function_warning_fill_loading_raw = exports.function_warning_fill_hover_raw = exports.function_warning_fill_default_raw = exports.function_warning_content_pressed_raw = exports.function_warning_content_loading_raw = exports.function_warning_content_hover_raw = exports.function_warning_content_default_raw = exports.function_warning_900_raw = exports.function_warning_800_raw = exports.function_warning_700_raw = exports.function_warning_600_raw = exports.function_warning_500_raw = exports.function_warning_50_raw = exports.function_warning_400_raw = exports.function_warning_350_raw = exports.function_warning_300_raw = exports.function_warning_200_raw = exports.function_warning_100_raw = exports.function_success_on_success_fill_raw = exports.function_success_fill_solid_03_raw = exports.function_success_fill_solid_02_raw = exports.function_success_fill_solid_01_raw = exports.function_success_fill_pressed_raw = exports.function_success_fill_loading_raw = exports.function_success_fill_hover_raw = exports.function_success_fill_default_raw = exports.function_success_content_pressed_raw = void 0;
exports.text_title_raw = exports.text_placeholder_raw = exports.text_link_pressed_raw = exports.text_link_normal_raw = exports.text_link_loading_raw = exports.text_link_hover_raw = exports.text_link_disabled_raw = exports.text_disabled_raw = exports.text_caption_raw = exports.static_white_raw = exports.static_white_pressed_raw = exports.static_white_hover_raw = exports.static_black_raw = exports.shadow_pri_raw = exports.shadow_default_raw = exports.primary_pri_900_raw = exports.primary_pri_800_raw = exports.primary_pri_700_raw = exports.primary_pri_600_raw = exports.primary_pri_500_raw = exports.primary_pri_50_raw = exports.primary_pri_400_raw = exports.primary_pri_350_raw = exports.primary_pri_300_raw = exports.primary_pri_200_raw = exports.primary_pri_100_raw = exports.primary_on_primary_fill_raw = exports.primary_fill_solid_03_raw = exports.primary_fill_solid_02_raw = exports.primary_fill_solid_01_raw = exports.primary_fill_pressed_raw = exports.primary_fill_loading_raw = exports.primary_fill_hover_raw = exports.primary_fill_default_raw = exports.primary_content_pressed_raw = exports.primary_content_loading_raw = exports.primary_content_hover_raw = exports.primary_content_default_raw = exports.line_border_component_raw = exports.line_border_card_raw = exports.inverse_primary_content_pressed_raw = exports.inverse_primary_content_hover_raw = exports.inverse_primary_content_default_raw = exports.illustration_yellow_raw = void 0;
// autogen by ../scripts/gen-ts-from-json.ts
exports.AI_cursor_raw = '94, 104, 232';
exports.AI_icon_in_button_raw = '123, 131, 247';
exports.B100_raw = '23, 49, 102';
exports.B200_raw = '25, 66, 148';
exports.B300_raw = '38, 85, 182';
exports.B350_raw = '39, 95, 206';
exports.B400_raw = '51, 112, 235';
exports.B50_raw = '21, 35, 64';
exports.B500_raw = '76, 136, 255';
exports.B600_raw = '117, 164, 255';
exports.B700_raw = '143, 180, 255';
exports.B800_raw = '189, 210, 255';
exports.B900_raw = '224, 233, 255';
exports.C100_raw = '89, 28, 63';
exports.C200_raw = '120, 43, 87';
exports.C300_raw = '148, 56, 108';
exports.C350_raw = '171, 65, 125';
exports.C400_raw = '194, 74, 142';
exports.C50_raw = '58, 24, 43';
exports.C500_raw = '219, 94, 164';
exports.C600_raw = '237, 119, 186';
exports.C700_raw = '252, 148, 207';
exports.C800_raw = '255, 194, 229';
exports.C900_raw = '255, 224, 242';
exports.G100_raw = '23, 59, 18';
exports.G200_raw = '33, 81, 26';
exports.G300_raw = '41, 102, 33';
exports.G350_raw = '47, 117, 38';
exports.G400_raw = '53, 135, 42';
exports.G50_raw = '14, 43, 10';
exports.G500_raw = '65, 158, 52';
exports.G600_raw = '81, 186, 67';
exports.G700_raw = '105, 204, 92';
exports.G800_raw = '153, 228, 144';
exports.G900_raw = '203, 245, 198';
exports.I100_raw = '42, 45, 105';
exports.I200_raw = '55, 61, 144';
exports.I300_raw = '71, 79, 184';
exports.I350_raw = '81, 90, 214';
exports.I400_raw = '94, 104, 232';
exports.I50_raw = '30, 32, 74';
exports.I500_raw = '123, 131, 247';
exports.I600_raw = '148, 153, 247';
exports.I700_raw = '170, 175, 248';
exports.I800_raw = '205, 208, 249';
exports.I900_raw = '231, 233, 254';
exports.L100_raw = '48, 56, 4';
exports.L200_raw = '64, 76, 6';
exports.L300_raw = '83, 97, 10';
exports.L350_raw = '95, 112, 10';
exports.L400_raw = '107, 127, 5';
exports.L50_raw = '33, 39, 2';
exports.L500_raw = '130, 153, 10';
exports.L600_raw = '147, 175, 4';
exports.L700_raw = '166, 195, 19';
exports.L800_raw = '194, 225, 45';
exports.L900_raw = '227, 243, 145';
exports.N00_raw = '10, 10, 10';
exports.N100_raw = '41, 41, 41';
exports.N1000_raw = '255, 255, 255';
exports.N200_raw = '55, 55, 55';
exports.N300_raw = '67, 67, 67';
exports.N350_raw = '80, 80, 80';
exports.N400_raw = '95, 95, 95';
exports.N50_raw = '26, 26, 26';
exports.N500_raw = '117, 117, 117';
exports.N600_raw = '166, 166, 166';
exports.N650_raw = '207, 207, 207';
exports.N700_raw = '224, 224, 224';
exports.N800_raw = '232, 232, 232';
exports.N900_raw = '235, 235, 235';
exports.N950_raw = '248, 248, 248';
exports.O100_raw = '74, 43, 16';
exports.O200_raw = '104, 58, 18';
exports.O300_raw = '138, 74, 25';
exports.O350_raw = '161, 83, 23';
exports.O400_raw = '184, 94, 26';
exports.O50_raw = '51, 33, 11';
exports.O500_raw = '219, 112, 24';
exports.O600_raw = '243, 135, 27';
exports.O700_raw = '248, 158, 68';
exports.O800_raw = '254, 200, 139';
exports.O900_raw = '254, 231, 205';
exports.P100_raw = '63, 32, 115';
exports.P200_raw = '85, 41, 163';
exports.P300_raw = '108, 57, 198';
exports.P350_raw = '124, 74, 212';
exports.P400_raw = '140, 85, 236';
exports.P50_raw = '43, 25, 74';
exports.P500_raw = '165, 117, 250';
exports.P600_raw = '184, 143, 254';
exports.P700_raw = '197, 163, 255';
exports.P800_raw = '219, 200, 253';
exports.P900_raw = '239, 229, 255';
exports.R100_raw = '89, 31, 29';
exports.R200_raw = '123, 37, 36';
exports.R300_raw = '160, 51, 49';
exports.R350_raw = '179, 58, 55';
exports.R400_raw = '209, 70, 66';
exports.R50_raw = '61, 26, 25';
exports.R500_raw = '240, 91, 86';
exports.R600_raw = '246, 130, 126';
exports.R700_raw = '248, 158, 155';
exports.R800_raw = '253, 198, 196';
exports.R900_raw = '254, 227, 226';
exports.S100_raw = '56, 50, 12';
exports.S200_raw = '87, 77, 1';
exports.S300_raw = '122, 106, 1';
exports.S350_raw = '156, 135, 2';
exports.S400_raw = '201, 178, 24';
exports.S50_raw = '41, 37, 10';
exports.S500_raw = '229, 205, 23';
exports.S600_raw = '245, 223, 54';
exports.S700_raw = '247, 230, 99';
exports.S800_raw = '250, 237, 122';
exports.S900_raw = '255, 245, 158';
exports.T100_raw = '23, 59, 54';
exports.T200_raw = '29, 78, 71';
exports.T300_raw = '37, 102, 94';
exports.T350_raw = '34, 115, 105';
exports.T400_raw = '25, 133, 120';
exports.T50_raw = '19, 41, 38';
exports.T500_raw = '31, 161, 143';
exports.T600_raw = '26, 183, 161';
exports.T700_raw = '23, 207, 181';
exports.T800_raw = '101, 231, 213';
exports.T900_raw = '183, 247, 239';
exports.V100_raw = '84, 24, 84';
exports.V200_raw = '113, 40, 113';
exports.V300_raw = '139, 55, 139';
exports.V350_raw = '164, 61, 164';
exports.V400_raw = '181, 74, 181';
exports.V50_raw = '59, 21, 59';
exports.V500_raw = '214, 97, 214';
exports.V600_raw = '225, 127, 225';
exports.V700_raw = '233, 155, 233';
exports.V800_raw = '244, 195, 244';
exports.V900_raw = '252, 223, 252';
exports.Vnext_bg_content_header_raw = '34, 34, 34';
exports.Vnext_bg_sub_navigation_raw = '30, 30, 30';
exports.Vnext_bg_thirdly_navigation_raw = '26, 26, 26';
exports.W100_raw = '16, 54, 71';
exports.W200_raw = '22, 67, 89';
exports.W300_raw = '19, 90, 120';
exports.W350_raw = '20, 108, 145';
exports.W400_raw = '26, 127, 171';
exports.W50_raw = '21, 40, 48';
exports.W500_raw = '16, 153, 204';
exports.W600_raw = '37, 178, 229';
exports.W700_raw = '81, 195, 240';
exports.W800_raw = '137, 223, 254';
exports.W900_raw = '199, 240, 255';
exports.Y100_raw = '71, 52, 9';
exports.Y200_raw = '99, 71, 15';
exports.Y300_raw = '138, 100, 25';
exports.Y350_raw = '173, 125, 21';
exports.Y400_raw = '212, 155, 11';
exports.Y50_raw = '48, 37, 10';
exports.Y500_raw = '240, 182, 34';
exports.Y600_raw = '251, 203, 70';
exports.Y700_raw = '252, 212, 86';
exports.Y800_raw = '255, 222, 117';
exports.Y900_raw = '255, 234, 163';
exports.bg_base_raw = '10, 10, 10';
exports.bg_body_overlay_raw = '41, 41, 41';
exports.bg_body_raw = '26, 26, 26';
exports.bg_content_base_raw = '18, 18, 18';
exports.bg_filler_raw = '55, 55, 55';
exports.bg_float_base_raw = '26, 26, 26';
exports.bg_float_overlay_raw = '55, 55, 55';
exports.bg_float_raw = '41, 41, 41';
exports.bg_mask_raw = '0, 0, 0';
exports.bg_pricolor_raw = '51, 112, 235';
exports.bg_sub_navigation_raw = '38, 38, 38';
exports.bg_tips_raw = '80, 80, 80';
exports.colorful_blue_raw = '76, 136, 255';
exports.colorful_carmine_raw = '194, 74, 142';
exports.colorful_green_raw = '65, 158, 52';
exports.colorful_indigo_raw = '94, 104, 232';
exports.colorful_lime_raw = '147, 175, 4';
exports.colorful_neutral_raw = '117, 117, 117';
exports.colorful_orange_raw = '243, 135, 27';
exports.colorful_purple_raw = '140, 85, 236';
exports.colorful_red_raw = '240, 91, 86';
exports.colorful_sunflower_raw = '229, 205, 23';
exports.colorful_turquoise_raw = '31, 161, 143';
exports.colorful_violet_raw = '181, 74, 181';
exports.colorful_wathet_raw = '16, 153, 204';
exports.colorful_yellow_raw = '240, 182, 34';
exports.fill_disabled_raw = '95, 95, 95';
exports.function_danger_100_raw = '89, 31, 29';
exports.function_danger_200_raw = '123, 37, 36';
exports.function_danger_300_raw = '160, 51, 49';
exports.function_danger_350_raw = '179, 58, 55';
exports.function_danger_400_raw = '209, 70, 66';
exports.function_danger_50_raw = '61, 26, 25';
exports.function_danger_500_raw = '240, 91, 86';
exports.function_danger_600_raw = '246, 130, 126';
exports.function_danger_700_raw = '248, 158, 155';
exports.function_danger_800_raw = '253, 198, 196';
exports.function_danger_900_raw = '254, 227, 226';
exports.function_danger_content_default_raw = '240, 91, 86';
exports.function_danger_content_hover_raw = '209, 70, 66';
exports.function_danger_content_loading_raw = '160, 51, 49';
exports.function_danger_content_pressed_raw = '246, 130, 126';
exports.function_danger_fill_default_raw = '209, 70, 66';
exports.function_danger_fill_hover_raw = '179, 58, 55';
exports.function_danger_fill_loading_raw = '160, 51, 49';
exports.function_danger_fill_pressed_raw = '240, 91, 86';
exports.function_danger_fill_solid_01_raw = '61, 26, 25';
exports.function_danger_fill_solid_02_raw = '89, 31, 29';
exports.function_danger_fill_solid_03_raw = '123, 37, 36';
exports.function_danger_on_danger_fill_raw = '255, 255, 255';
exports.function_info_100_raw = '23, 49, 102';
exports.function_info_200_raw = '25, 66, 148';
exports.function_info_300_raw = '38, 85, 182';
exports.function_info_350_raw = '39, 95, 206';
exports.function_info_400_raw = '51, 112, 235';
exports.function_info_50_raw = '21, 35, 64';
exports.function_info_500_raw = '76, 136, 255';
exports.function_info_600_raw = '117, 164, 255';
exports.function_info_700_raw = '143, 180, 255';
exports.function_info_800_raw = '189, 210, 255';
exports.function_info_900_raw = '224, 233, 255';
exports.function_info_content_default_raw = '76, 136, 255';
exports.function_info_content_hover_raw = '51, 112, 235';
exports.function_info_content_loading_raw = '38, 85, 182';
exports.function_info_content_pressed_raw = '117, 164, 255';
exports.function_info_fill_default_raw = '51, 112, 235';
exports.function_info_fill_hover_raw = '39, 95, 206';
exports.function_info_fill_loading_raw = '38, 85, 182';
exports.function_info_fill_pressed_raw = '76, 136, 255';
exports.function_info_fill_solid_01_raw = '21, 35, 64';
exports.function_info_fill_solid_02_raw = '23, 49, 102';
exports.function_info_fill_solid_03_raw = '25, 66, 148';
exports.function_info_on_info_fill_raw = '255, 255, 255';
exports.function_success_100_raw = '23, 59, 18';
exports.function_success_200_raw = '33, 81, 26';
exports.function_success_300_raw = '41, 102, 33';
exports.function_success_350_raw = '47, 117, 38';
exports.function_success_400_raw = '53, 135, 42';
exports.function_success_50_raw = '14, 43, 10';
exports.function_success_500_raw = '65, 158, 52';
exports.function_success_600_raw = '81, 186, 67';
exports.function_success_700_raw = '105, 204, 92';
exports.function_success_800_raw = '153, 228, 144';
exports.function_success_900_raw = '203, 245, 198';
exports.function_success_content_default_raw = '81, 186, 67';
exports.function_success_content_hover_raw = '65, 158, 52';
exports.function_success_content_loading_raw = '41, 102, 33';
exports.function_success_content_pressed_raw = '105, 204, 92';
exports.function_success_fill_default_raw = '65, 158, 52';
exports.function_success_fill_hover_raw = '53, 135, 42';
exports.function_success_fill_loading_raw = '41, 102, 33';
exports.function_success_fill_pressed_raw = '81, 186, 67';
exports.function_success_fill_solid_01_raw = '14, 43, 10';
exports.function_success_fill_solid_02_raw = '23, 59, 18';
exports.function_success_fill_solid_03_raw = '33, 81, 26';
exports.function_success_on_success_fill_raw = '255, 255, 255';
exports.function_warning_100_raw = '74, 43, 16';
exports.function_warning_200_raw = '104, 58, 18';
exports.function_warning_300_raw = '138, 74, 25';
exports.function_warning_350_raw = '161, 83, 23';
exports.function_warning_400_raw = '184, 94, 26';
exports.function_warning_50_raw = '51, 33, 11';
exports.function_warning_500_raw = '219, 112, 24';
exports.function_warning_600_raw = '243, 135, 27';
exports.function_warning_700_raw = '248, 158, 68';
exports.function_warning_800_raw = '254, 200, 139';
exports.function_warning_900_raw = '254, 231, 205';
exports.function_warning_content_default_raw = '243, 135, 27';
exports.function_warning_content_hover_raw = '219, 112, 24';
exports.function_warning_content_loading_raw = '138, 74, 25';
exports.function_warning_content_pressed_raw = '248, 158, 68';
exports.function_warning_fill_default_raw = '243, 135, 27';
exports.function_warning_fill_hover_raw = '219, 112, 24';
exports.function_warning_fill_loading_raw = '138, 74, 25';
exports.function_warning_fill_pressed_raw = '248, 158, 68';
exports.function_warning_fill_solid_01_raw = '51, 33, 11';
exports.function_warning_fill_solid_02_raw = '74, 43, 16';
exports.function_warning_fill_solid_03_raw = '104, 58, 18';
exports.function_warning_on_warning_fill_raw = '255, 255, 255';
exports.icon_disabled_raw = '95, 95, 95';
exports.icon_n1_raw = '232, 232, 232';
exports.icon_n2_raw = '166, 166, 166';
exports.icon_n3_raw = '117, 117, 117';
exports.illustration_blue_a_raw = '224, 224, 224';
exports.illustration_blue_b_raw = '255, 255, 255';
exports.illustration_blue_c_raw = '95, 95, 95';
exports.illustration_blue_d_raw = '38, 85, 182';
exports.illustration_blue_e_raw = '76, 136, 255';
exports.illustration_neutral_a_raw = '235, 235, 235';
exports.illustration_neutral_b_raw = '55, 55, 55';
exports.illustration_neutral_d_raw = '255, 255, 255';
exports.illustration_orange_raw = '243, 135, 27';
exports.illustration_red_raw = '240, 91, 86';
exports.illustration_static_black_raw = '0, 0, 0';
exports.illustration_static_white_raw = '255, 255, 255';
exports.illustration_turquoise_raw = '23, 207, 181';
exports.illustration_wathet_raw = '81, 195, 240';
exports.illustration_yellow_raw = '240, 182, 34';
exports.inverse_primary_content_default_raw = '117, 164, 255';
exports.inverse_primary_content_hover_raw = '143, 180, 255';
exports.inverse_primary_content_pressed_raw = '76, 136, 255';
exports.line_border_card_raw = '235, 235, 235';
exports.line_border_component_raw = '80, 80, 80';
exports.primary_content_default_raw = '76, 136, 255';
exports.primary_content_hover_raw = '51, 112, 235';
exports.primary_content_loading_raw = '38, 85, 182';
exports.primary_content_pressed_raw = '117, 164, 255';
exports.primary_fill_default_raw = '51, 112, 235';
exports.primary_fill_hover_raw = '39, 95, 206';
exports.primary_fill_loading_raw = '38, 85, 182';
exports.primary_fill_pressed_raw = '76, 136, 255';
exports.primary_fill_solid_01_raw = '21, 35, 64';
exports.primary_fill_solid_02_raw = '23, 49, 102';
exports.primary_fill_solid_03_raw = '25, 66, 148';
exports.primary_on_primary_fill_raw = '255, 255, 255';
exports.primary_pri_100_raw = '23, 49, 102';
exports.primary_pri_200_raw = '25, 66, 148';
exports.primary_pri_300_raw = '38, 85, 182';
exports.primary_pri_350_raw = '39, 95, 206';
exports.primary_pri_400_raw = '51, 112, 235';
exports.primary_pri_50_raw = '21, 35, 64';
exports.primary_pri_500_raw = '76, 136, 255';
exports.primary_pri_600_raw = '117, 164, 255';
exports.primary_pri_700_raw = '143, 180, 255';
exports.primary_pri_800_raw = '189, 210, 255';
exports.primary_pri_900_raw = '224, 233, 255';
exports.shadow_default_raw = '0, 0, 0';
exports.shadow_pri_raw = '36, 91, 219';
exports.static_black_raw = '0, 0, 0';
exports.static_white_hover_raw = '255, 255, 255';
exports.static_white_pressed_raw = '255, 255, 255';
exports.static_white_raw = '255, 255, 255';
exports.text_caption_raw = '166, 166, 166';
exports.text_disabled_raw = '95, 95, 95';
exports.text_link_disabled_raw = '95, 95, 95';
exports.text_link_hover_raw = '51, 112, 235';
exports.text_link_loading_raw = '38, 85, 182';
exports.text_link_normal_raw = '76, 136, 255';
exports.text_link_pressed_raw = '117, 164, 255';
exports.text_placeholder_raw = '117, 117, 117';
exports.text_title_raw = '235, 235, 235';
