// @universe-design/styles/cjs/preset-light.atom-raw.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.L350_raw = exports.L300_raw = exports.L200_raw = exports.L100_raw = exports.I900_raw = exports.I800_raw = exports.I700_raw = exports.I600_raw = exports.I500_raw = exports.I50_raw = exports.I400_raw = exports.I350_raw = exports.I300_raw = exports.I200_raw = exports.I100_raw = exports.G900_raw = exports.G800_raw = exports.G700_raw = exports.G600_raw = exports.G500_raw = exports.G50_raw = exports.G400_raw = exports.G350_raw = exports.G300_raw = exports.G200_raw = exports.G100_raw = exports.C900_raw = exports.C800_raw = exports.C700_raw = exports.C600_raw = exports.C500_raw = exports.C50_raw = exports.C400_raw = exports.C350_raw = exports.C300_raw = exports.C200_raw = exports.C100_raw = exports.B900_raw = exports.B800_raw = exports.B700_raw = exports.B600_raw = exports.B500_raw = exports.B50_raw = exports.B400_raw = exports.B350_raw = exports.B300_raw = exports.B200_raw = exports.B100_raw = exports.AI_icon_in_button_raw = exports.AI_cursor_raw = void 0;
exports.R50_raw = exports.R400_raw = exports.R350_raw = exports.R300_raw = exports.R200_raw = exports.R100_raw = exports.P900_raw = exports.P800_raw = exports.P700_raw = exports.P600_raw = exports.P500_raw = exports.P50_raw = exports.P400_raw = exports.P350_raw = exports.P300_raw = exports.P200_raw = exports.P100_raw = exports.O900_raw = exports.O800_raw = exports.O700_raw = exports.O600_raw = exports.O500_raw = exports.O50_raw = exports.O400_raw = exports.O350_raw = exports.O300_raw = exports.O200_raw = exports.O100_raw = exports.N950_raw = exports.N900_raw = exports.N800_raw = exports.N700_raw = exports.N650_raw = exports.N600_raw = exports.N500_raw = exports.N50_raw = exports.N400_raw = exports.N350_raw = exports.N300_raw = exports.N200_raw = exports.N1000_raw = exports.N100_raw = exports.N00_raw = exports.L900_raw = exports.L800_raw = exports.L700_raw = exports.L600_raw = exports.L500_raw = exports.L50_raw = exports.L400_raw = void 0;
exports.W700_raw = exports.W600_raw = exports.W500_raw = exports.W50_raw = exports.W400_raw = exports.W350_raw = exports.W300_raw = exports.W200_raw = exports.W100_raw = exports.Vnext_bg_thirdly_navigation_raw = exports.Vnext_bg_sub_navigation_raw = exports.Vnext_bg_content_header_raw = exports.V900_raw = exports.V800_raw = exports.V700_raw = exports.V600_raw = exports.V500_raw = exports.V50_raw = exports.V400_raw = exports.V350_raw = exports.V300_raw = exports.V200_raw = exports.V100_raw = exports.T900_raw = exports.T800_raw = exports.T700_raw = exports.T600_raw = exports.T500_raw = exports.T50_raw = exports.T400_raw = exports.T350_raw = exports.T300_raw = exports.T200_raw = exports.T100_raw = exports.S900_raw = exports.S800_raw = exports.S700_raw = exports.S600_raw = exports.S500_raw = exports.S50_raw = exports.S400_raw = exports.S350_raw = exports.S300_raw = exports.S200_raw = exports.S100_raw = exports.R900_raw = exports.R800_raw = exports.R700_raw = exports.R600_raw = exports.R500_raw = void 0;
exports.function_danger_800_raw = exports.function_danger_700_raw = exports.function_danger_600_raw = exports.function_danger_500_raw = exports.function_danger_50_raw = exports.function_danger_400_raw = exports.function_danger_350_raw = exports.function_danger_300_raw = exports.function_danger_200_raw = exports.function_danger_100_raw = exports.fill_disabled_raw = exports.colorful_yellow_raw = exports.colorful_wathet_raw = exports.colorful_violet_raw = exports.colorful_turquoise_raw = exports.colorful_sunflower_raw = exports.colorful_red_raw = exports.colorful_purple_raw = exports.colorful_orange_raw = exports.colorful_neutral_raw = exports.colorful_lime_raw = exports.colorful_indigo_raw = exports.colorful_green_raw = exports.colorful_carmine_raw = exports.colorful_blue_raw = exports.bg_tips_raw = exports.bg_sub_navigation_raw = exports.bg_pricolor_raw = exports.bg_mask_raw = exports.bg_float_raw = exports.bg_float_overlay_raw = exports.bg_float_base_raw = exports.bg_filler_raw = exports.bg_content_base_raw = exports.bg_body_raw = exports.bg_body_overlay_raw = exports.bg_base_raw = exports.Y900_raw = exports.Y800_raw = exports.Y700_raw = exports.Y600_raw = exports.Y500_raw = exports.Y50_raw = exports.Y400_raw = exports.Y350_raw = exports.Y300_raw = exports.Y200_raw = exports.Y100_raw = exports.W900_raw = exports.W800_raw = void 0;
exports.function_success_content_loading_raw = exports.function_success_content_hover_raw = exports.function_success_content_default_raw = exports.function_success_900_raw = exports.function_success_800_raw = exports.function_success_700_raw = exports.function_success_600_raw = exports.function_success_500_raw = exports.function_success_50_raw = exports.function_success_400_raw = exports.function_success_350_raw = exports.function_success_300_raw = exports.function_success_200_raw = exports.function_success_100_raw = exports.function_info_on_info_fill_raw = exports.function_info_fill_solid_03_raw = exports.function_info_fill_solid_02_raw = exports.function_info_fill_solid_01_raw = exports.function_info_fill_pressed_raw = exports.function_info_fill_loading_raw = exports.function_info_fill_hover_raw = exports.function_info_fill_default_raw = exports.function_info_content_pressed_raw = exports.function_info_content_loading_raw = exports.function_info_content_hover_raw = exports.function_info_content_default_raw = exports.function_info_900_raw = exports.function_info_800_raw = exports.function_info_700_raw = exports.function_info_600_raw = exports.function_info_500_raw = exports.function_info_50_raw = exports.function_info_400_raw = exports.function_info_350_raw = exports.function_info_300_raw = exports.function_info_200_raw = exports.function_info_100_raw = exports.function_danger_on_danger_fill_raw = exports.function_danger_fill_solid_03_raw = exports.function_danger_fill_solid_02_raw = exports.function_danger_fill_solid_01_raw = exports.function_danger_fill_pressed_raw = exports.function_danger_fill_loading_raw = exports.function_danger_fill_hover_raw = exports.function_danger_fill_default_raw = exports.function_danger_content_pressed_raw = exports.function_danger_content_loading_raw = exports.function_danger_content_hover_raw = exports.function_danger_content_default_raw = exports.function_danger_900_raw = void 0;
exports.illustration_wathet_raw = exports.illustration_turquoise_raw = exports.illustration_static_white_raw = exports.illustration_static_black_raw = exports.illustration_red_raw = exports.illustration_orange_raw = exports.illustration_neutral_d_raw = exports.illustration_neutral_b_raw = exports.illustration_neutral_a_raw = exports.illustration_blue_e_raw = exports.illustration_blue_d_raw = exports.illustration_blue_c_raw = exports.illustration_blue_b_raw = exports.illustration_blue_a_raw = exports.icon_n3_raw = exports.icon_n2_raw = exports.icon_n1_raw = exports.icon_disabled_raw = exports.function_warning_on_warning_fill_raw = exports.function_warning_fill_solid_03_raw = exports.function_warning_fill_solid_02_raw = exports.function_warning_fill_solid_01_raw = exports.function_warning_fill_pressed_raw = exports.function_warning_fill_loading_raw = exports.function_warning_fill_hover_raw = exports.function_warning_fill_default_raw = exports.function_warning_content_pressed_raw = exports.function_warning_content_loading_raw = exports.function_warning_content_hover_raw = exports.function_warning_content_default_raw = exports.function_warning_900_raw = exports.function_warning_800_raw = exports.function_warning_700_raw = exports.function_warning_600_raw = exports.function_warning_500_raw = exports.function_warning_50_raw = exports.function_warning_400_raw = exports.function_warning_350_raw = exports.function_warning_300_raw = exports.function_warning_200_raw = exports.function_warning_100_raw = exports.function_success_on_success_fill_raw = exports.function_success_fill_solid_03_raw = exports.function_success_fill_solid_02_raw = exports.function_success_fill_solid_01_raw = exports.function_success_fill_pressed_raw = exports.function_success_fill_loading_raw = exports.function_success_fill_hover_raw = exports.function_success_fill_default_raw = exports.function_success_content_pressed_raw = void 0;
exports.text_title_raw = exports.text_placeholder_raw = exports.text_link_pressed_raw = exports.text_link_normal_raw = exports.text_link_loading_raw = exports.text_link_hover_raw = exports.text_link_disabled_raw = exports.text_disabled_raw = exports.text_caption_raw = exports.static_white_raw = exports.static_white_pressed_raw = exports.static_white_hover_raw = exports.static_black_raw = exports.shadow_pri_raw = exports.shadow_default_raw = exports.primary_pri_900_raw = exports.primary_pri_800_raw = exports.primary_pri_700_raw = exports.primary_pri_600_raw = exports.primary_pri_500_raw = exports.primary_pri_50_raw = exports.primary_pri_400_raw = exports.primary_pri_350_raw = exports.primary_pri_300_raw = exports.primary_pri_200_raw = exports.primary_pri_100_raw = exports.primary_on_primary_fill_raw = exports.primary_fill_solid_03_raw = exports.primary_fill_solid_02_raw = exports.primary_fill_solid_01_raw = exports.primary_fill_pressed_raw = exports.primary_fill_loading_raw = exports.primary_fill_hover_raw = exports.primary_fill_default_raw = exports.primary_content_pressed_raw = exports.primary_content_loading_raw = exports.primary_content_hover_raw = exports.primary_content_default_raw = exports.line_border_component_raw = exports.line_border_card_raw = exports.inverse_primary_content_pressed_raw = exports.inverse_primary_content_hover_raw = exports.inverse_primary_content_default_raw = exports.illustration_yellow_raw = void 0;
// autogen by ../scripts/gen-ts-from-json.ts
exports.AI_cursor_raw = '71, 82, 230';
exports.AI_icon_in_button_raw = '71, 82, 230';
exports.B100_raw = '224, 233, 255';
exports.B200_raw = '194, 212, 255';
exports.B300_raw = '148, 180, 255';
exports.B350_raw = '122, 162, 255';
exports.B400_raw = '80, 131, 251';
exports.B50_raw = '240, 244, 255';
exports.B500_raw = '51, 109, 244';
exports.B600_raw = '20, 86, 240';
exports.B700_raw = '4, 66, 210';
exports.B800_raw = '0, 47, 158';
exports.B900_raw = '0, 34, 112';
exports.C100_raw = '254, 226, 242';
exports.C200_raw = '248, 196, 225';
exports.C300_raw = '245, 152, 204';
exports.C350_raw = '235, 120, 184';
exports.C400_raw = '223, 88, 165';
exports.C50_raw = '254, 240, 248';
exports.C500_raw = '204, 57, 140';
exports.C600_raw = '184, 40, 121';
exports.C700_raw = '157, 21, 98';
exports.C800_raw = '115, 7, 68';
exports.C900_raw = '85, 12, 53';
exports.G100_raw = '208, 245, 206';
exports.G200_raw = '149, 229, 153';
exports.G300_raw = '92, 209, 104';
exports.G350_raw = '53, 189, 75';
exports.G400_raw = '50, 166, 69';
exports.G50_raw = '228, 250, 225';
exports.G500_raw = '37, 136, 50';
exports.G600_raw = '26, 117, 38';
exports.G700_raw = '11, 96, 23';
exports.G800_raw = '4, 67, 12';
exports.G900_raw = '2, 44, 7';
exports.I100_raw = '233, 234, 251';
exports.I200_raw = '204, 207, 249';
exports.I300_raw = '171, 176, 242';
exports.I350_raw = '149, 155, 238';
exports.I400_raw = '117, 125, 240';
exports.I50_raw = '242, 243, 253';
exports.I500_raw = '91, 101, 245';
exports.I600_raw = '71, 82, 230';
exports.I700_raw = '51, 61, 204';
exports.I800_raw = '30, 39, 164';
exports.I900_raw = '21, 27, 112';
exports.L100_raw = '227, 240, 163';
exports.L200_raw = '200, 221, 95';
exports.L300_raw = '162, 193, 11';
exports.L350_raw = '145, 173, 0';
exports.L400_raw = '123, 146, 7';
exports.L50_raw = '242, 248, 211';
exports.L500_raw = '107, 127, 6';
exports.L600_raw = '92, 109, 8';
exports.L700_raw = '74, 88, 4';
exports.L800_raw = '51, 61, 0';
exports.L900_raw = '38, 46, 0';
exports.N00_raw = '255, 255, 255';
exports.N100_raw = '242, 243, 245';
exports.N1000_raw = '0, 0, 0';
exports.N200_raw = '239, 240, 241';
exports.N300_raw = '222, 224, 227';
exports.N350_raw = '208, 211, 214';
exports.N400_raw = '187, 191, 196';
exports.N50_raw = '245, 246, 247';
exports.N500_raw = '143, 149, 158';
exports.N600_raw = '100, 106, 115';
exports.N650_raw = '81, 86, 93';
exports.N700_raw = '55, 60, 67';
exports.N800_raw = '43, 47, 54';
exports.N900_raw = '31, 35, 41';
exports.N950_raw = '15, 17, 20';
exports.O100_raw = '254, 231, 205';
exports.O200_raw = '254, 196, 139';
exports.O300_raw = '255, 157, 76';
exports.O350_raw = '255, 129, 26';
exports.O400_raw = '237, 109, 12';
exports.O50_raw = '255, 243, 229';
exports.O500_raw = '194, 87, 5';
exports.O600_raw = '164, 73, 4';
exports.O700_raw = '133, 58, 5';
exports.O800_raw = '100, 43, 2';
exports.O900_raw = '59, 26, 2';
exports.P100_raw = '239, 230, 254';
exports.P200_raw = '220, 201, 253';
exports.P300_raw = '200, 169, 252';
exports.P350_raw = '183, 145, 250';
exports.P400_raw = '159, 111, 241';
exports.P50_raw = '245, 240, 254';
exports.P500_raw = '141, 85, 237';
exports.P600_raw = '122, 53, 240';
exports.P700_raw = '97, 31, 214';
exports.P800_raw = '72, 17, 166';
exports.P900_raw = '47, 0, 128';
exports.R100_raw = '254, 227, 226';
exports.R200_raw = '253, 198, 196';
exports.R300_raw = '248, 158, 155';
exports.R350_raw = '255, 117, 112';
exports.R400_raw = '245, 74, 69';
exports.R50_raw = '254, 240, 240';
exports.R500_raw = '226, 46, 40';
exports.R600_raw = '192, 42, 38';
exports.R700_raw = '161, 28, 23';
exports.R800_raw = '116, 25, 21';
exports.R900_raw = '89, 6, 3';
exports.S100_raw = '255, 252, 163';
exports.S200_raw = '255, 246, 122';
exports.S300_raw = '255, 242, 88';
exports.S350_raw = '255, 233, 40';
exports.S400_raw = '229, 206, 0';
exports.S50_raw = '255, 255, 219';
exports.S500_raw = '194, 168, 0';
exports.S600_raw = '143, 124, 0';
exports.S700_raw = '92, 79, 0';
exports.S800_raw = '66, 55, 0';
exports.S900_raw = '44, 37, 2';
exports.T100_raw = '196, 242, 236';
exports.T200_raw = '111, 232, 216';
exports.T300_raw = '51, 214, 192';
exports.T350_raw = '45, 190, 171';
exports.T400_raw = '16, 168, 147';
exports.T50_raw = '226, 248, 245';
exports.T500_raw = '15, 133, 117';
exports.T600_raw = '6, 112, 98';
exports.T700_raw = '4, 93, 81';
exports.T800_raw = '3, 68, 59';
exports.T900_raw = '2, 49, 42';
exports.V100_raw = '249, 226, 249';
exports.V200_raw = '243, 196, 243';
exports.V300_raw = '229, 156, 229';
exports.V350_raw = '222, 129, 222';
exports.V400_raw = '207, 94, 207';
exports.V50_raw = '252, 238, 252';
exports.V500_raw = '191, 61, 191';
exports.V600_raw = '166, 48, 166';
exports.V700_raw = '135, 39, 135';
exports.V800_raw = '106, 17, 106';
exports.V900_raw = '82, 0, 82';
exports.Vnext_bg_content_header_raw = '252, 252, 252';
exports.Vnext_bg_sub_navigation_raw = '245, 245, 245';
exports.Vnext_bg_thirdly_navigation_raw = '252, 252, 252';
exports.W100_raw = '202, 239, 252';
exports.W200_raw = '151, 220, 252';
exports.W300_raw = '62, 195, 247';
exports.W350_raw = '37, 176, 231';
exports.W400_raw = '18, 149, 202';
exports.W50_raw = '231, 248, 254';
exports.W500_raw = '4, 127, 176';
exports.W600_raw = '7, 106, 148';
exports.W700_raw = '15, 88, 122';
exports.W800_raw = '6, 65, 92';
exports.W900_raw = '7, 43, 61';
exports.Y100_raw = '250, 237, 194';
exports.Y200_raw = '252, 223, 126';
exports.Y300_raw = '250, 211, 85';
exports.Y350_raw = '255, 198, 10';
exports.Y400_raw = '217, 153, 4';
exports.Y50_raw = '251, 244, 223';
exports.Y500_raw = '173, 122, 3';
exports.Y600_raw = '134, 91, 3';
exports.Y700_raw = '111, 74, 1';
exports.Y800_raw = '87, 54, 1';
exports.Y900_raw = '56, 34, 1';
exports.bg_base_raw = '242, 243, 245';
exports.bg_body_overlay_raw = '245, 246, 247';
exports.bg_body_raw = '255, 255, 255';
exports.bg_content_base_raw = '248, 249, 250';
exports.bg_filler_raw = '239, 240, 241';
exports.bg_float_base_raw = '242, 243, 245';
exports.bg_float_overlay_raw = '245, 246, 247';
exports.bg_float_raw = '255, 255, 255';
exports.bg_mask_raw = '0, 0, 0';
exports.bg_pricolor_raw = '80, 131, 251';
exports.bg_sub_navigation_raw = '245, 246, 247';
exports.bg_tips_raw = '31, 35, 41';
exports.colorful_blue_raw = '51, 109, 244';
exports.colorful_carmine_raw = '223, 88, 165';
exports.colorful_green_raw = '53, 189, 75';
exports.colorful_indigo_raw = '91, 101, 245';
exports.colorful_lime_raw = '145, 173, 0';
exports.colorful_neutral_raw = '143, 149, 158';
exports.colorful_orange_raw = '255, 129, 26';
exports.colorful_purple_raw = '141, 85, 237';
exports.colorful_red_raw = '245, 74, 69';
exports.colorful_sunflower_raw = '255, 233, 40';
exports.colorful_turquoise_raw = '45, 190, 171';
exports.colorful_violet_raw = '191, 61, 191';
exports.colorful_wathet_raw = '37, 176, 231';
exports.colorful_yellow_raw = '255, 198, 10';
exports.fill_disabled_raw = '187, 191, 196';
exports.function_danger_100_raw = '254, 227, 226';
exports.function_danger_200_raw = '253, 198, 196';
exports.function_danger_300_raw = '248, 158, 155';
exports.function_danger_350_raw = '255, 117, 112';
exports.function_danger_400_raw = '245, 74, 69';
exports.function_danger_50_raw = '254, 240, 240';
exports.function_danger_500_raw = '226, 46, 40';
exports.function_danger_600_raw = '192, 42, 38';
exports.function_danger_700_raw = '161, 28, 23';
exports.function_danger_800_raw = '116, 25, 21';
exports.function_danger_900_raw = '89, 6, 3';
exports.function_danger_content_default_raw = '245, 74, 69';
exports.function_danger_content_hover_raw = '255, 117, 112';
exports.function_danger_content_loading_raw = '248, 158, 155';
exports.function_danger_content_pressed_raw = '226, 46, 40';
exports.function_danger_fill_default_raw = '245, 74, 69';
exports.function_danger_fill_hover_raw = '255, 117, 112';
exports.function_danger_fill_loading_raw = '248, 158, 155';
exports.function_danger_fill_pressed_raw = '226, 46, 40';
exports.function_danger_fill_solid_01_raw = '254, 240, 240';
exports.function_danger_fill_solid_02_raw = '254, 227, 226';
exports.function_danger_fill_solid_03_raw = '253, 198, 196';
exports.function_danger_on_danger_fill_raw = '255, 255, 255';
exports.function_info_100_raw = '224, 233, 255';
exports.function_info_200_raw = '194, 212, 255';
exports.function_info_300_raw = '148, 180, 255';
exports.function_info_350_raw = '122, 162, 255';
exports.function_info_400_raw = '80, 131, 251';
exports.function_info_50_raw = '240, 244, 255';
exports.function_info_500_raw = '51, 109, 244';
exports.function_info_600_raw = '20, 86, 240';
exports.function_info_700_raw = '4, 66, 210';
exports.function_info_800_raw = '0, 47, 158';
exports.function_info_900_raw = '0, 34, 112';
exports.function_info_content_default_raw = '20, 86, 240';
exports.function_info_content_hover_raw = '51, 109, 244';
exports.function_info_content_loading_raw = '148, 180, 255';
exports.function_info_content_pressed_raw = '4, 66, 210';
exports.function_info_fill_default_raw = '20, 86, 240';
exports.function_info_fill_hover_raw = '51, 109, 244';
exports.function_info_fill_loading_raw = '148, 180, 255';
exports.function_info_fill_pressed_raw = '4, 66, 210';
exports.function_info_fill_solid_01_raw = '240, 244, 255';
exports.function_info_fill_solid_02_raw = '224, 233, 255';
exports.function_info_fill_solid_03_raw = '194, 212, 255';
exports.function_info_on_info_fill_raw = '255, 255, 255';
exports.function_success_100_raw = '208, 245, 206';
exports.function_success_200_raw = '149, 229, 153';
exports.function_success_300_raw = '92, 209, 104';
exports.function_success_350_raw = '53, 189, 75';
exports.function_success_400_raw = '50, 166, 69';
exports.function_success_50_raw = '228, 250, 225';
exports.function_success_500_raw = '37, 136, 50';
exports.function_success_600_raw = '26, 117, 38';
exports.function_success_700_raw = '11, 96, 23';
exports.function_success_800_raw = '4, 67, 12';
exports.function_success_900_raw = '2, 44, 7';
exports.function_success_content_default_raw = '50, 166, 69';
exports.function_success_content_hover_raw = '53, 189, 75';
exports.function_success_content_loading_raw = '92, 209, 104';
exports.function_success_content_pressed_raw = '37, 136, 50';
exports.function_success_fill_default_raw = '53, 189, 75';
exports.function_success_fill_hover_raw = '92, 209, 104';
exports.function_success_fill_loading_raw = '92, 209, 104';
exports.function_success_fill_pressed_raw = '50, 166, 69';
exports.function_success_fill_solid_01_raw = '228, 250, 225';
exports.function_success_fill_solid_02_raw = '208, 245, 206';
exports.function_success_fill_solid_03_raw = '149, 229, 153';
exports.function_success_on_success_fill_raw = '255, 255, 255';
exports.function_warning_100_raw = '254, 231, 205';
exports.function_warning_200_raw = '254, 196, 139';
exports.function_warning_300_raw = '255, 157, 76';
exports.function_warning_350_raw = '255, 129, 26';
exports.function_warning_400_raw = '237, 109, 12';
exports.function_warning_50_raw = '255, 243, 229';
exports.function_warning_500_raw = '194, 87, 5';
exports.function_warning_600_raw = '164, 73, 4';
exports.function_warning_700_raw = '133, 58, 5';
exports.function_warning_800_raw = '100, 43, 2';
exports.function_warning_900_raw = '59, 26, 2';
exports.function_warning_content_default_raw = '237, 109, 12';
exports.function_warning_content_hover_raw = '255, 129, 26';
exports.function_warning_content_loading_raw = '255, 157, 76';
exports.function_warning_content_pressed_raw = '194, 87, 5';
exports.function_warning_fill_default_raw = '255, 129, 26';
exports.function_warning_fill_hover_raw = '255, 157, 76';
exports.function_warning_fill_loading_raw = '255, 157, 76';
exports.function_warning_fill_pressed_raw = '237, 109, 12';
exports.function_warning_fill_solid_01_raw = '255, 243, 229';
exports.function_warning_fill_solid_02_raw = '254, 231, 205';
exports.function_warning_fill_solid_03_raw = '254, 196, 139';
exports.function_warning_on_warning_fill_raw = '255, 255, 255';
exports.icon_disabled_raw = '187, 191, 196';
exports.icon_n1_raw = '43, 47, 54';
exports.icon_n2_raw = '100, 106, 115';
exports.icon_n3_raw = '143, 149, 158';
exports.illustration_blue_a_raw = '0, 34, 112';
exports.illustration_blue_b_raw = '0, 34, 112';
exports.illustration_blue_c_raw = '0, 34, 112';
exports.illustration_blue_d_raw = '0, 34, 112';
exports.illustration_blue_e_raw = '51, 109, 244';
exports.illustration_neutral_a_raw = '255, 255, 255';
exports.illustration_neutral_b_raw = '255, 255, 255';
exports.illustration_neutral_d_raw = '143, 149, 158';
exports.illustration_orange_raw = '255, 129, 26';
exports.illustration_red_raw = '245, 74, 69';
exports.illustration_static_black_raw = '0, 0, 0';
exports.illustration_static_white_raw = '255, 255, 255';
exports.illustration_turquoise_raw = '51, 214, 192';
exports.illustration_wathet_raw = '62, 195, 247';
exports.illustration_yellow_raw = '255, 198, 10';
exports.inverse_primary_content_default_raw = '80, 131, 251';
exports.inverse_primary_content_hover_raw = '122, 162, 255';
exports.inverse_primary_content_pressed_raw = '51, 109, 244';
exports.line_border_card_raw = '222, 224, 227';
exports.line_border_component_raw = '208, 211, 214';
exports.primary_content_default_raw = '20, 86, 240';
exports.primary_content_hover_raw = '51, 109, 244';
exports.primary_content_loading_raw = '148, 180, 255';
exports.primary_content_pressed_raw = '4, 66, 210';
exports.primary_fill_default_raw = '20, 86, 240';
exports.primary_fill_hover_raw = '51, 109, 244';
exports.primary_fill_loading_raw = '148, 180, 255';
exports.primary_fill_pressed_raw = '4, 66, 210';
exports.primary_fill_solid_01_raw = '240, 244, 255';
exports.primary_fill_solid_02_raw = '224, 233, 255';
exports.primary_fill_solid_03_raw = '194, 212, 255';
exports.primary_on_primary_fill_raw = '255, 255, 255';
exports.primary_pri_100_raw = '224, 233, 255';
exports.primary_pri_200_raw = '194, 212, 255';
exports.primary_pri_300_raw = '148, 180, 255';
exports.primary_pri_350_raw = '122, 162, 255';
exports.primary_pri_400_raw = '80, 131, 251';
exports.primary_pri_50_raw = '240, 244, 255';
exports.primary_pri_500_raw = '51, 109, 244';
exports.primary_pri_600_raw = '20, 86, 240';
exports.primary_pri_700_raw = '4, 66, 210';
exports.primary_pri_800_raw = '0, 47, 158';
exports.primary_pri_900_raw = '0, 34, 112';
exports.shadow_default_raw = '31, 35, 41';
exports.shadow_pri_raw = '20, 86, 240';
exports.static_black_raw = '0, 0, 0';
exports.static_white_hover_raw = '255, 255, 255';
exports.static_white_pressed_raw = '255, 255, 255';
exports.static_white_raw = '255, 255, 255';
exports.text_caption_raw = '100, 106, 115';
exports.text_disabled_raw = '187, 191, 196';
exports.text_link_disabled_raw = '187, 191, 196';
exports.text_link_hover_raw = '51, 109, 244';
exports.text_link_loading_raw = '148, 180, 255';
exports.text_link_normal_raw = '20, 86, 240';
exports.text_link_pressed_raw = '4, 66, 210';
exports.text_placeholder_raw = '143, 149, 158';
exports.text_title_raw = '31, 35, 41';
