// @universe-design/react/es/skeleton/ShapeItem.js
import clsx from 'clsx';
import merge from 'lodash/merge';
import React from 'react';
import { useCssScope } from '../config-provider';
import './style';
const ShapeItem = props => {
  const {
    className,
    typeName,
    style
  } = props;
  const cssScope = useCssScope('skeleton__item');
  return /*#__PURE__*/React.createElement("div", {
    className: clsx([cssScope(), cssScope(`-${typeName}`), className]),
    style: style
  }, /*#__PURE__*/React.createElement("div", {
    className: cssScope('__animation-layer')
  }));
};
export const SkeletonContent = props => {
  const {
    className,
    style,
    width,
    height
  } = props;
  // skip undefined
  const mergedStyle = merge(style, {
    width,
    height
  });
  return /*#__PURE__*/React.createElement(ShapeItem, {
    typeName: "content",
    className: className,
    style: mergedStyle
  });
};
export const SkeletonAvatar = props => {
  const {
    className,
    style,
    size
  } = props;
  // skip undefined
  const mergedStyle = merge(style, {
    width: size,
    height: size
  });
  return /*#__PURE__*/React.createElement(ShapeItem, {
    typeName: "avatar",
    className: className,
    style: mergedStyle
  });
};
export const SkeletonParagraph = props => {
  const {
    className,
    style,
    rows = 3
  } = props;
  const cssScope = useCssScope('skeleton');
  return /*#__PURE__*/React.createElement("div", {
    className: clsx([cssScope('__paragraph'), className]),
    style: style
  }, [...Array(rows)].map((_, index) => /*#__PURE__*/React.createElement(SkeletonContent, {
    key: index
  })));
};
