// @byted-service/env/lib/deprecated.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isPerfMode = exports.getIPFamily = exports.getIPVersion = exports.getPodEnv = exports.isDpsAuthEnabled = exports.getRPCServiceMeshIngressAddr = exports.isHTTPServiceMeshIngressEnabled = exports.isRPCServiceMeshIngressEnabled = exports.getRPCServiceMeshEgressAddr = exports.isRPCServiceMeshEgressEnabled = exports.getMongoServiceMeshAddr = exports.isMongoServiceMeshEnabled = exports.getMysqlServiceMeshAddr = exports.isMysqlServiceMeshEnabled = exports.isFaas = exports.getPPEEnv = exports.getGatewayByEnv = exports.getGateway = exports.isDocker = void 0;
const child_process_1 = require("child_process");
const nodejs_1 = require("./nodejs");
/** @deprecated use env.isInTCE() instead */
const isDocker = () => {
    return (process.env.IS_DOCKER_ENV === 'true' ||
        process.env.IS_TCE_DOCKER_ENV === '1');
};
exports.isDocker = isDocker;
/** @deprecated please implement it yourself */
const getGateway = () => {
    // gulu bootstrap.sh 里面定义了 ROUTE_IP
    if (process.env.ROUTE_IP) {
        return process.env.ROUTE_IP;
    }
    if ((0, exports.isDocker)()) {
        if (process.env.IS_HOST_NETWORK) {
            return '127.0.0.1';
        }
        // TCE 实例上，优先使用 MY_HOST_IP 获取，`ip route` 的方式不准确
        if (process.env.MY_HOST_IP) {
            return process.env.MY_HOST_IP;
        }
        const pattern = /default via (\S+)/;
        try {
            const route = ((0, child_process_1.execSync)('/opt/tiger/consul_deploy/bin/determine_dc.sh', {
                encoding: 'utf-8',
            }) || '').trim();
            const match = route.match(pattern);
            if (match) {
                return match[1];
            }
        }
        catch (err) {
            // IGNORE
        }
    }
    return '127.0.0.1';
};
exports.getGateway = getGateway;
/** @deprecated please implement it yourself */
const getGatewayByEnv = () => {
    if (process.env.ROUTE_IP) {
        // 不管什么环境，只要定义了 ROUTE_IP 优先使用这个环境变量
        // 对于 isProd() || isTest() 的时候，没有影响，因为 getGateway 的默认逻辑就是
        // 判断 ROUTE_IP，所以影响的是本地开发环境。
        return process.env.ROUTE_IP;
    }
    /* istanbul ignore next */
    return (0, nodejs_1.isProd)() || (0, nodejs_1.isTest)() ? (0, exports.getGateway)() : '127.0.0.1';
};
exports.getGatewayByEnv = getGatewayByEnv;
/** @deprecated use env.getEnv() instead */
const getPPEEnv = () => {
    return process.env.SERVICE_ENV || '';
};
exports.getPPEEnv = getPPEEnv;
/** @deprecated please implement it yourself */
const isFaas = () => {
    return process.env.IS_FAAS_ENV === 'True';
};
exports.isFaas = isFaas;
/** @deprecated please implement it yourself */
const isMysqlServiceMeshEnabled = () => {
    // BOE 和 TCE 环境变量名一致
    return process.env.TCE_ENABLE_MYSQL_SIDECAR_EGRESS === 'True'
        ? true
        : false;
};
exports.isMysqlServiceMeshEnabled = isMysqlServiceMeshEnabled;
/** @deprecated please implement it yourself */
const getMysqlServiceMeshAddr = () => {
    return (0, exports.isMysqlServiceMeshEnabled)()
        ? process.env.SERVICE_MESH_MYSQL_ADDR
        : '-';
};
exports.getMysqlServiceMeshAddr = getMysqlServiceMeshAddr;
/** @deprecated please implement it yourself */
const isMongoServiceMeshEnabled = () => {
    // BOE 和 TCE 环境变量名一致
    return process.env.TCE_ENABLE_MONGO_SIDECAR_EGRESS === 'True';
};
exports.isMongoServiceMeshEnabled = isMongoServiceMeshEnabled;
/** @deprecated please implement it yourself */
const getMongoServiceMeshAddr = () => {
    return (0, exports.isMongoServiceMeshEnabled)()
        ? process.env.SERVICE_MESH_MONGO_ADDR
        : '-';
};
exports.getMongoServiceMeshAddr = getMongoServiceMeshAddr;
/** @deprecated please implement it yourself */
const isRPCServiceMeshEgressEnabled = () => {
    // BOE 和 TCE 环境变量名一致
    return process.env.TCE_ENABLE_SIDECAR_EGRESS === 'True' ? true : false;
};
exports.isRPCServiceMeshEgressEnabled = isRPCServiceMeshEgressEnabled;
/** @deprecated please implement it yourself */
const getRPCServiceMeshEgressAddr = () => {
    return (0, exports.isRPCServiceMeshEgressEnabled)()
        ? process.env.SERVICE_MESH_EGRESS_ADDR
        : '-';
};
exports.getRPCServiceMeshEgressAddr = getRPCServiceMeshEgressAddr;
/** @deprecated please implement it yourself */
const isRPCServiceMeshIngressEnabled = () => {
    // BOE 和 TCE 环境变量名一致
    return process.env.TCE_ENABLE_SIDECAR_INGRESS === 'True' ? true : false;
};
exports.isRPCServiceMeshIngressEnabled = isRPCServiceMeshIngressEnabled;
/** @deprecated please implement it yourself */
const isHTTPServiceMeshIngressEnabled = () => {
    // BOE 和 TCE 环境变量名一致
    return process.env.TCE_ENABLE_HTTP_SIDECAR_INGRESS === 'True'
        ? true
        : false;
};
exports.isHTTPServiceMeshIngressEnabled = isHTTPServiceMeshIngressEnabled;
/** @deprecated please implement it yourself */
const getRPCServiceMeshIngressAddr = () => {
    return (0, exports.isRPCServiceMeshIngressEnabled)()
        ? process.env.SERVICE_MESH_INGRESS_ADDR
        : '-';
};
exports.getRPCServiceMeshIngressAddr = getRPCServiceMeshIngressAddr;
/** @deprecated please implement it yourself */
const isDpsAuthEnabled = () => {
    return process.env.SEC_KV_AUTH === '1';
};
exports.isDpsAuthEnabled = isDpsAuthEnabled;
/** @deprecated use env.getEnv() instead */
const getPodEnv = () => {
    return process.env.TCE_ENV || 'prod';
};
exports.getPodEnv = getPodEnv;
/** @deprecated please implement it yourself */
const getIPVersion = () => {
    if (process.env.MY_HOST_IP && process.env.MY_HOST_IPV6) {
        return 'dual-stack';
    }
    if (process.env.MY_HOST_IPV6) {
        return 'IPv6';
    }
    if (process.env.MY_HOST_IP) {
        return 'IPv4';
    }
    return '';
};
exports.getIPVersion = getIPVersion;
/** @deprecated please implement it yourself */
exports.getIPFamily = exports.getIPVersion;
/** @deprecated please implement it yourself */
const isPerfMode = () => {
    return Number(process.env.PERF_MODE) === 1;
};
exports.isPerfMode = isPerfMode;
