// @byted-service/env/lib/idc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setIpv6File = exports.setIDCFile = exports.setIDC = exports.getIDCFromHost = exports.getIDC = exports.init = exports.IDC = void 0;
const tslib_1 = require("tslib");
const fs_1 = tslib_1.__importDefault(require("fs"));
const child_process_1 = require("child_process");
const util_1 = require("util");
const ip_1 = require("./internal/net/ip");
const search_1 = require("./internal/sort/search");
const idc_util_1 = require("./idc_util");
const region_1 = require("./region");
const readFileSync = fs_1.default.readFileSync;
const readFileAsync = (0, util_1.promisify)(fs_1.default.readFile);
/* istanbul ignore next */
const noop = () => {
    //
};
var idc_util_2 = require("./idc_util");
Object.defineProperty(exports, "IDC", { enumerable: true, get: function () { return idc_util_2.IDC; } });
const agentFileV4 = '/opt/tmp/consul_agent/netsegment';
const localFileV4 = '/opt/tiger/chadc/netsegment';
const agentFileV6 = '/opt/tmp/consul_agent/netsegment6';
const localFileV6 = '/opt/tiger/chadc/netsegment6';
const idcRefreshDur = 5 * 60 * 1000; // 5min
let idc = ''; // local idc (string)
let ipv4File = ''; // file name (string)
let ipv6File = ''; // file name (string)
let ipv4Segments = []; // ipv4 segments - idc ([]*ipv4Segment)
let ipv6Segments = []; // ipv6 segments - idc ([]*ipv6Segment)
let idcRefreshIntervalId = null;
const init = () => {
    idc = '';
    ipv4File = '';
    ipv6File = '';
    ipv4Segments = [];
    ipv6Segments = [];
    if (!idcRefreshIntervalId) {
        refreshIDCs();
    }
    refreshIDCsSync();
};
exports.init = init;
/**
 * IDC .
 */
const getIDC = () => {
    if (idc) {
        return idc;
    }
    if (process.env.RUNTIME_IDC_NAME) {
        idc = process.env.RUNTIME_IDC_NAME;
        return idc;
    }
    try {
        const output = readFileSync('/opt/tmp/consul_agent/datacenter', {
            encoding: 'utf-8',
        }).trim();
        if (output) {
            idc = output;
            return idc;
        }
    }
    catch (err) {
        // do nothing
    }
    try {
        const output = ((0, child_process_1.execSync)('/opt/tiger/consul_deploy/bin/determine_dc.sh', {
            encoding: 'utf-8',
            stdio: [null, null, null],
        }) || '').trim();
        if ((0, region_1.hasIDC)(output)) {
            idc = output;
            return idc;
        }
    }
    catch (err) {
        // do nothing
    }
    try {
        const output = ((0, child_process_1.execSync)('sd report | grep "Data center" | awk \'{print $3}\'', {
            encoding: 'utf-8',
            stdio: [null, null, null],
        }) || '').trim();
        if ((0, region_1.hasIDC)(output)) {
            idc = output;
            return idc;
        }
    }
    catch (err) {
        // do nothing
    }
    idc = idc_util_1.IDC.UnknownIDC;
    return idc;
};
exports.getIDC = getIDC;
/**
 * GetIDCFromHost .
 */
const getIDCFromHost = (ip) => {
    const netIP = (0, ip_1.parseIP)(ip);
    // ipv4
    if (ip.includes('.') && netIP) {
        return lookupIpv4(netIP);
    }
    // ipv6
    if (ip.includes(':') && netIP) {
        return lookupIpv6(netIP);
    }
    return idc_util_1.IDC.UnknownIDC;
};
exports.getIDCFromHost = getIDCFromHost;
/**
 * SetIDC .
 */
const setIDC = (v) => {
    idc = v;
};
exports.setIDC = setIDC;
/**
 * SetIDCFile only replace file ipv4 segment
 */
const setIDCFile = (file) => {
    ipv4File = file;
    updateIpv4SegmentsSync();
};
exports.setIDCFile = setIDCFile;
/**
 * SetIpv6File only replace file ipv6 segment6
 */
const setIpv6File = (file) => {
    ipv6File = file;
    updateIpv6SegmentsSync();
};
exports.setIpv6File = setIpv6File;
/**
 * ATTENTION: IT COMES WITH A LOOP, DON'T CALL IT AGAIN.
 */
const refreshIDCs = () => {
    idcRefreshIntervalId = setTimeout(refreshIDCs, idcRefreshDur).unref();
    updateIpv4Segments().catch(noop);
    updateIpv6Segments().catch(noop);
};
/**
 * updateIpv4Segments
 */
const updateIpv4Segments = async () => {
    const ipv4s = [];
    const addItem = (ipNet, idc) => {
        ipv4s.push(newIpv4Segment(ipNet, idc));
    };
    const fileV4s = [agentFileV4, localFileV4];
    await readSegments(ipv4File, fileV4s, addItem);
    // sort & merge
    ipv4Segments = mergeIpv4Segment(ipv4s);
};
/**
 * updateIpv6Segments
 */
const updateIpv6Segments = async () => {
    const ipv6s = [];
    const addItem = (ipNet, idc) => {
        ipv6s.push(newIpv6Segment(ipNet, idc));
    };
    const fileV6s = [agentFileV6, localFileV6];
    await readSegments(ipv6File, fileV6s, addItem);
    // sort & merge
    ipv6Segments = mergeIpv6Segment(ipv6s);
};
/**
 * readSegments add all read segments
 */
const readSegments = async (userFile, defaultFiles, addItem) => {
    let lines = [];
    // read user's setting first
    const file = userFile;
    if (file) {
        lines = await readFile(file);
    }
    // if no user's setting, read default files
    if (lines.length === 0) {
        for (let i = 0; i < defaultFiles.length; i++) {
            const file = defaultFiles[i];
            lines = lines.concat(await readFile(file));
        }
    }
    // Deduplication: key=segment, value=idc
    const lineMap = {};
    for (let i = 0; i < lines.length; i++) {
        const line = lines[i].trim();
        const s = line.split(' ');
        if (s.length !== 2) {
            continue;
        }
        // The former line has higher priority
        const [seg, idc] = s;
        if (lineMap[seg]) {
            continue;
        }
        const [, ipNet, err] = (0, ip_1.parseCIDR)(seg);
        if (err || !ipNet) {
            continue;
        }
        // add item
        addItem(ipNet, idc);
        lineMap[seg] = idc;
    }
};
/**
 * readFile return lines
 */
const readFile = async (file) => {
    try {
        const content = await readFileAsync(file, { encoding: 'utf-8' });
        return content.split('\n');
    }
    catch (err) {
        /* istanbul ignore next */
        return [];
    }
};
/* ipv4/ipv6 归属IDC 查询算法说明
 *
 * 本文件查询采用了 基于网段分片的二分查找算法
 * 我们知道 对于形如 10.10.8.0/22 对应的 ipv4 网段为 [10.10.8.0, 10.10.11.255], 换算为 uint32 为 [0xA0A0800, 0xA0A0BFF]
 * 因此, 对于给定的一组网段信息, 均可以转换为 uint32 的闭区间 形如 [a, b],[b+1,c],[d,e][e+1,f]....
 * 我们对上述闭区间从小到大排序, 则对于给定的 ipv4 地址, 则可以先转为 uint32, 在通过二分查找得到其网段归属, 获得该网段的 IDC 信息
 * ipv6 同理
 *
 * 算法效率
 * 基于二分查找的时间复杂度为 O(log(n)), 空间复杂度 O(n) 对于头条目前网段配置(ipv4 网段个数 60-70), 该算法最多查询 7 次
 * 附: 为什么不采用前缀树匹配 ?
 * 前缀树实际会很大, 如 10.10.8.0/22 有 22 位前缀, 如果再加上形如 10.224.0.0/18, 10.224.128.0/17, 10.11.0.0/16 的网段,
 * 则即使最优存储情况下, 前缀树需要保留 16-22 位(实际更多) 的分叉节点, 树的层级远大于 7, 并且存储结构比有序数组复杂, 因此不是最优选择
 */
// ipv4 section
/**
 * lookupIpv4 .
 */
const lookupIpv4 = (ip) => {
    if (ip.length === ip_1.IPv6len) {
        ip = ip.slice(12, 16);
    }
    if (ip.length !== ip_1.IPv4len) {
        return idc_util_1.IDC.UnknownIDC;
    }
    const ipv4s = ipv4Segments;
    if (ipv4s.length === 0) {
        return idc_util_1.IDC.UnknownIDC;
    }
    // Binary search
    const target = bigEndianUint32(ip);
    const _search = (i) => {
        return target <= ipv4s[i].end;
    };
    const index = (0, search_1.search)(ipv4s.length, _search);
    // search not found
    if (index < 0 || index >= ipv4s.length) {
        return idc_util_1.IDC.UnknownIDC;
    }
    /* istanbul ignore next */
    if (ipv4s[index].start <= target && target <= ipv4s[index].end) {
        return ipv4s[index].idc;
    }
    else {
        return idc_util_1.IDC.UnknownIDC;
    }
};
/**
 * newIpv4Segment .
 */
const newIpv4Segment = (ipNet, idc) => {
    const segment = {
        idc,
        start: bigEndianUint32(ipNet.IP),
    };
    segment.end = segment.start + (bigEndianUint32(ipNet.Mask) ^ 0xffffffff);
    return segment;
};
/* ipv4/ipv6 网段合并算法说明
 *
 * 由于网段配置文件采用了最长匹配原则, 形如 10.8.0.0/16, 10.8.0.0/17, 存在包含关系, 需要打平为互不包含的网段列表。
 * 我们先对原始网段配置, 按起始位置从小到大排序, 对于起始相同的, 先结束的小网段排在后;
 * 再遍历有序网段列表, 对于前后网段有包含关系的, 拆分成互不包含的网段, 由此得到打平后的有序网段配置。
 */
/**
 * mergeIpv4Segment Merge adjacent segments if idc is same
 */
const mergeIpv4Segment = (ipv4s) => {
    const merged = [];
    if (ipv4s.length === 0) {
        return merged;
    }
    /* istanbul ignore next */
    ipv4s.sort((a, b) => {
        if (a.start < b.start) {
            return -1;
        }
        // 起始位置一样, 大网段在前
        if (a.start === b.start && a.end > b.end) {
            return -1;
        }
        return a.end === b.end ? 0 : 1;
    });
    // 先打平所有网段
    let flatten = [ipv4s[0]];
    let former;
    let latter;
    let formerIdx = 0;
    let latterIdx = 1;
    for (; latterIdx < ipv4s.length;) {
        former = flatten[formerIdx];
        latter = ipv4s[latterIdx];
        // 判断前后两网段是否交叉, former.end >= latter.start
        if (former.end >= latter.start) {
            const tmp = [];
            latterIdx++;
            if (former.start === latter.start) {
                tmp.push(latter);
            }
            else {
                tmp.push({
                    idc: former.idc,
                    start: former.start,
                    end: latter.start - 1,
                }, latter);
            }
            // former.end 更后
            if (former.end > latter.end) {
                former.start = latter.end + 1;
                tmp.push(former);
            }
            if (formerIdx + 1 !== flatten.length) {
                tmp.push(...flatten.slice(formerIdx + 1));
            }
            flatten = flatten.slice(0, formerIdx);
            flatten.push(...tmp);
        }
        else {
            // 两网段相互独立
            formerIdx++;
            if (formerIdx === flatten.length) {
                latterIdx++;
                flatten.push(latter);
            }
        }
    }
    // 合并相邻网段
    former = flatten[0];
    merged.push(flatten[0]);
    for (let i = 1; i < flatten.length; i++) {
        latter = flatten[i];
        // idc 相同, 并且 former.end+1 == latter.start
        if (former.idc === latter.idc && former.end + 1 === latter.start) {
            former.end = latter.end;
        }
        else {
            former = latter;
            merged.push(latter);
        }
    }
    return merged;
};
// ipv6 section
/**
 * lookupIpv6 .
 */
const lookupIpv6 = (ip) => {
    if (ip.length !== ip_1.IPv6len) {
        return idc_util_1.IDC.UnknownIDC;
    }
    const ipv6s = ipv6Segments;
    if (ipv6s.length === 0) {
        return idc_util_1.IDC.UnknownIDC;
    }
    // Binary search
    const target = bigEndianUint128(ip);
    const _search = (i) => {
        return leU128(target, ipv6s[i].end);
    };
    const index = (0, search_1.search)(ipv6s.length, _search);
    // search not found
    if (index < 0 || index >= ipv6s.length) {
        return idc_util_1.IDC.UnknownIDC;
    }
    /* istanbul ignore next */
    if (leU128(ipv6s[index].start, target) &&
        leU128(target, ipv6s[index].end)) {
        return ipv6s[index].idc;
    }
    else {
        return idc_util_1.IDC.UnknownIDC;
    }
};
/**
 * newIpv6Segment .
 */
const newIpv6Segment = (ipNet, idc) => {
    const segment = {
        idc,
        start: bigEndianUint128(ipNet.IP),
    };
    segment.end = sumU128(segment.start, negU128(bigEndianUint128(ipNet.Mask)));
    return segment;
};
/**
 * ipv6 地址遵循前缀匹配原则, 因此先基于网段起始位置排序, 在根据包含关系划分网段
 */
const mergeIpv6Segment = (ipv6s) => {
    const merged = [];
    if (ipv6s.length === 0) {
        return merged;
    }
    /* istanbul ignore next */
    ipv6s.sort((a, b) => {
        if (ltU128(a.start, b.start)) {
            return -1;
        }
        // 起始位置一样, 大网段在前
        if (eqU128(a.start, b.start) && gtU128(a.end, b.end)) {
            return -1;
        }
        return eqU128(a.end, b.end) ? 0 : 1;
    });
    // 打平所有网段
    let flatten = [ipv6s[0]];
    let former;
    let latter;
    let formerIdx = 0;
    let latterIdx = 1;
    for (; latterIdx < ipv6s.length;) {
        former = flatten[formerIdx];
        latter = ipv6s[latterIdx];
        // 判断前后两网段是否交叉, former.end >= latter.start
        if (geU128(former.end, latter.start)) {
            const tmp = [];
            latterIdx++;
            if (eqU128(former.start, latter.start)) {
                tmp.push(latter);
            }
            else {
                tmp.push({
                    idc: former.idc,
                    start: former.start,
                    end: diffU128(latter.start, [0, 0, 0, 1]),
                }, latter);
            }
            // former.end 更后
            /* istanbul ignore next */
            if (gtU128(former.end, latter.end)) {
                former.start = sumU128(latter.end, [0, 0, 0, 1]);
                tmp.push(former);
            }
            if (formerIdx + 1 !== flatten.length) {
                tmp.push(...flatten.slice(formerIdx + 1));
            }
            flatten = flatten.slice(0, formerIdx);
            flatten.push(...tmp);
        }
        else {
            // 两网段相互独立
            formerIdx++;
            if (formerIdx === flatten.length) {
                latterIdx++;
                flatten.push(latter);
            }
        }
    }
    // 合并相邻网段
    former = flatten[0];
    merged.push(flatten[0]);
    for (let i = 1; i < flatten.length; i++) {
        latter = flatten[i];
        // idc 相同, 并且 former.end+1 == latter.start
        if (former.idc === latter.idc &&
            eqU128(sumU128(former.end, [0, 0, 0, 1]), latter.start)) {
            former.end = latter.end;
        }
        else {
            former = latter;
            merged.push(latter);
        }
    }
    return merged;
};
/**
 * geU128 return true when i >= j.
 */
const geU128 = (i, j) => {
    return !ltU128(i, j);
};
/**
 * leU128 return true when i <= j.
 */
const leU128 = (i, j) => {
    return !ltU128(j, i);
};
/**
 * eqU128 return true when i == j.
 */
const eqU128 = (i, j) => {
    return i[0] === j[0] && i[1] === j[1] && i[2] === j[2] && i[3] === j[3];
};
/**
 * gtU128 return true when i > j.
 */
const gtU128 = (i, j) => {
    return ltU128(j, i);
};
/**
 * ltU128 return true when i < j.
 */
const ltU128 = (i, j) => {
    if (i[0] !== j[0]) {
        return i[0] < j[0];
    }
    if (i[1] !== j[1]) {
        return i[1] < j[1];
    }
    if (i[2] !== j[2]) {
        return i[2] < j[2];
    }
    return i[3] < j[3];
};
/**
 * sumU128 .
 */
const sumU128 = (i, j) => {
    const res = [0, 0, 0, 0];
    for (let k = 3; k > 0; k--) {
        const tmp = i[k] + j[k] + res[k];
        if (tmp > 0xffffffff) {
            res[k] = tmp % 0x100000000;
            res[k - 1] = (tmp - res[k]) / 0x100000000;
        }
        else {
            res[k] = tmp;
        }
    }
    res[0] += i[0] + j[0];
    return res;
};
/**
 * diffU128 .
 */
const diffU128 = (i, j) => {
    let res = [0, 0, 0, 0];
    res[0] = i[0];
    res[1] = i[1];
    res[2] = i[2];
    res[3] = i[3];
    for (let k = 3; k > 0; k--) {
        if (res[k] >= j[k]) {
            res[k] -= j[k];
        }
        else {
            const tmp = [0, 0, 0, 0];
            tmp[k - 1] = 1;
            res = diffU128(res, tmp);
            res[k] = 0x100000000 + res[k] - j[k];
        }
    }
    res[0] -= j[0];
    return res;
};
/**
 * negU128 Negate number.
 */
const negU128 = (i) => {
    return [
        (i[0] ^ 0xffffffff) >>> 0,
        (i[1] ^ 0xffffffff) >>> 0,
        (i[2] ^ 0xffffffff) >>> 0,
        (i[3] ^ 0xffffffff) >>> 0,
    ];
};
/**
 * bigEndianUint32 .
 */
const bigEndianUint32 = (p) => {
    return ((p[0] << 24) | (p[1] << 16) | (p[2] << 8) | p[3]) >>> 0;
};
/**
 * bigEndianUint128 .
 */
const bigEndianUint128 = (p) => {
    return [
        bigEndianUint32(p.slice(0, 4)),
        bigEndianUint32(p.slice(4, 8)),
        bigEndianUint32(p.slice(8, 12)),
        bigEndianUint32(p.slice(12, 16)),
    ];
};
// 同步刷新，保证 env 可用
const refreshIDCsSync = () => {
    updateIpv4SegmentsSync();
    updateIpv6SegmentsSync();
};
const updateIpv4SegmentsSync = () => {
    const ipv4s = [];
    const addItem = (ipNet, idc) => {
        ipv4s.push(newIpv4Segment(ipNet, idc));
    };
    const fileV4s = [agentFileV4, localFileV4];
    readSegmentsSync(ipv4File, fileV4s, addItem);
    // sort & merge
    ipv4Segments = mergeIpv4Segment(ipv4s);
};
const updateIpv6SegmentsSync = () => {
    const ipv6s = [];
    const addItem = (ipNet, idc) => {
        ipv6s.push(newIpv6Segment(ipNet, idc));
    };
    const fileV6s = [agentFileV6, localFileV6];
    readSegmentsSync(ipv6File, fileV6s, addItem);
    // sort & merge
    ipv6Segments = mergeIpv6Segment(ipv6s);
};
const readSegmentsSync = (userFile, defaultFiles, addItem) => {
    let lines = [];
    // read user's setting first
    const file = userFile;
    if (file) {
        lines = readFileSync_(file);
    }
    // if no user's setting, read default files
    if (lines.length === 0) {
        for (let i = 0; i < defaultFiles.length; i++) {
            const file = defaultFiles[i];
            lines = lines.concat(readFileSync_(file));
        }
    }
    // Deduplication: key=segment, value=idc
    const lineMap = {};
    for (let i = 0; i < lines.length; i++) {
        const line = lines[i].trim();
        const s = line.split(' ');
        if (s.length !== 2) {
            continue;
        }
        // The former line has higher priority
        const [seg, idc] = s;
        if (lineMap[seg]) {
            continue;
        }
        const [, ipNet, err] = (0, ip_1.parseCIDR)(seg);
        if (err || !ipNet) {
            continue;
        }
        // add item
        addItem(ipNet, idc);
        lineMap[seg] = idc;
    }
};
const readFileSync_ = (file) => {
    try {
        const content = readFileSync(file, { encoding: 'utf-8' });
        return content.split('\n');
    }
    catch (err) {
        return [];
    }
};
