// @byted-service/env/lib/port.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TCEPortRegistered = exports.TCEDebugPortRegistered = exports.TCEServicePortRegistered = exports.getTCEDebugPort = exports.getTCEServicePort = exports.init = void 0;
let tceServicePort = '';
let tceDebugPort = '';
const MaxPort = 65535;
const MinPort = 0;
const portMap = {};
const init = () => {
    tceServicePort = '';
    tceDebugPort = '';
    parseTCEPorts();
};
exports.init = init;
/**
 * TCEServicePort
 */
const getTCEServicePort = () => {
    return tceServicePort;
};
exports.getTCEServicePort = getTCEServicePort;
/**
 * TCEDebugPort
 */
const getTCEDebugPort = () => {
    return tceDebugPort;
};
exports.getTCEDebugPort = getTCEDebugPort;
const isValidPort = (portStr) => {
    const port = Number(portStr);
    return port >= MinPort && port <= MaxPort;
};
// TCEServicePortRegistered 返回 TCE_PRIMARY_PORT 对应的访问端口
const TCEServicePortRegistered = () => {
    return (0, exports.TCEPortRegistered)((0, exports.getTCEServicePort)());
};
exports.TCEServicePortRegistered = TCEServicePortRegistered;
// TCEDebugPortRegistered 返回 TCE_DEBUG_PORT 对应的访问端口
const TCEDebugPortRegistered = () => {
    return (0, exports.TCEPortRegistered)((0, exports.getTCEDebugPort)());
};
exports.TCEDebugPortRegistered = TCEDebugPortRegistered;
// TCEPortRegistered 返回服务端口对应的访问端口
const TCEPortRegistered = (svr) => {
    return portMap[svr] || '';
};
exports.TCEPortRegistered = TCEPortRegistered;
// https://bytedance.feishu.cn/wiki/wikcnR7f17phbl2h7ZVnj4KgRfg
// TCE 平台上用户注册所有端口的列表，通过空格分隔，按照顺序依次为 PORT0 PORT1 ... PORTN
const parseTCEPorts = () => {
    const servicePortStr = process.env.TCE_PRIMARY_PORT || '';
    if (isValidPort(servicePortStr)) {
        tceServicePort = servicePortStr;
    }
    const debugPortStr = process.env.TCE_DEBUG_PORT || '';
    if (isValidPort(debugPortStr)) {
        tceDebugPort = debugPortStr;
    }
    let raw = process.env.TCE_PORTS || '';
    if (raw === '') {
        return;
    }
    raw = raw.trim();
    const portList = raw.split(' ');
    portList.forEach((port, i) => {
        if (!isValidPort(port)) {
            return;
        }
        // TCE 平台往 Consul 注册的端口值，该变量下，PORT0会注册为主端口，其他的PORT1、2、... N，会注册到 Consul 的 Tag 里，如果有映射，这是经过端口映射后的值。
        const hostPort = process.env[`PORT${i}`] || '';
        if (!isValidPort(hostPort)) {
            return;
        }
        portMap[port] = hostPort;
    });
};
