// @byted-service/env/lib/product.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isProduct = exports.isEdgeEnvironment = exports.isPPE = exports.isBoeXfl = exports.isBoeTTP = exports.isBoeI18N = exports.isBoeCN = exports.isBoe = exports.isTesting = exports.init = void 0;
const tslib_1 = require("tslib");
const os_1 = tslib_1.__importDefault(require("os"));
const fs_1 = tslib_1.__importDefault(require("fs"));
const util_1 = require("util");
const idc_1 = require("./idc");
let isTesting_ = false;
let isPPE_ = false;
let isEdge = false;
let isProd = null;
const init = () => {
    isTesting_ = process.env.TESTING_PREFIX === 'offline';
    isPPE_ = process.env.TCE_HOST_ENV === 'ppe';
    isEdge = process.env.IS_EDGE_ENVIRONMENT === '1';
    isProd = null;
};
exports.init = init;
/**
 * IsTesting if in testing env
 */
const isTesting = () => {
    return isTesting_;
};
exports.isTesting = isTesting;
/**
 * IsBoe return true if it's in BOE by the IDC name
 */
const isBoe = () => {
    const idc = (0, idc_1.getIDC)();
    return (idc === 'boe' ||
        idc === 'boei18n' ||
        idc === 'boexfl' ||
        idc === 'boettp');
};
exports.isBoe = isBoe;
/**
 * IsBoeCN return true if the idc is in China
 */
const isBoeCN = () => {
    const idc = (0, idc_1.getIDC)();
    return idc === 'boe';
};
exports.isBoeCN = isBoeCN;
/**
 * IsBoeI18N return true if the idc is boei18n
 */
const isBoeI18N = () => {
    const idc = (0, idc_1.getIDC)();
    return idc === 'boei18n';
};
exports.isBoeI18N = isBoeI18N;
/**
 * IsBoeTTP return true if the idc is boettp
 */
const isBoeTTP = () => {
    const idc = (0, idc_1.getIDC)();
    return idc === 'boettp';
};
exports.isBoeTTP = isBoeTTP;
/**
 * IsBoeXfl return true if the idc is volcano boexfl
 */
const isBoeXfl = () => {
    const idc = (0, idc_1.getIDC)();
    return idc === 'boexfl';
};
exports.isBoeXfl = isBoeXfl;
const isPPE = () => {
    return isPPE_;
};
exports.isPPE = isPPE;
const isEdgeEnvironment = () => {
    return isEdge;
};
exports.isEdgeEnvironment = isEdgeEnvironment;
const processNameFormat = '/proc/%d/comm';
/**
 * IsProduct return true if current service is running on product enviroment else false
 */
const isProduct = () => {
    if (typeof isProd === 'boolean') {
        return isProd;
    }
    if ((0, exports.isTesting)() || (0, exports.isBoe)()) {
        isProd = false;
        return isProd;
    }
    // please see: https://wiki.bytedance.net/pages/viewpage.action?pageId=63229064
    if (process.env.IS_PROD_RUNTIME) {
        isProd = true;
        return isProd;
    }
    if (process.env.SERVICE_ENV) {
        isProd = true;
        return isProd;
    }
    const pn = parentProcName();
    if (!pn) {
        isProd = false;
        return isProd;
    }
    const u = os_1.default.userInfo();
    if (u.username === 'tiger' && (pn === 'supervise' || pn === 'systemd')) {
        isProd = true;
        return isProd;
    }
    isProd = false;
    return isProd;
};
exports.isProduct = isProduct;
/**
 * father's service name
 */
const parentProcName = () => {
    const ppid = process.ppid;
    try {
        const bs = fs_1.default.readFileSync((0, util_1.format)(processNameFormat, ppid), {
            encoding: 'utf-8',
        });
        return bs.trim();
    }
    catch (err) {
        return '';
    }
};
