// @byted-service/env/lib/region.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getIDCList = exports.idcList = exports.hasIDC = exports.setRegionFile = exports.getRegionFromIDC = exports.getRegion = exports.init = exports.REGION = void 0;
const tslib_1 = require("tslib");
const fs_1 = tslib_1.__importDefault(require("fs"));
const util_1 = require("util");
const idc_util_1 = require("./idc_util");
const readFileSync = fs_1.default.readFileSync;
const readFileAsync = (0, util_1.promisify)(fs_1.default.readFile);
/* istanbul ignore next */
const noop = () => {
    //
};
exports.REGION = {
    // UnknownRegion .
    UnknownRegion: '-',
    R_CN: 'CN',
    R_SG: 'SG',
    R_US: 'US',
    R_MALIVA: 'MALIVA',
    R_ALISG: 'ALISG',
    R_CA: 'CA',
    R_BOE: 'BOE',
    R_ChinaBOE2: 'China-BOE2',
    R_SUITEKA: 'SUITEKA',
    R_SUITEKA2: 'SUITEKA2',
    R_SUITEKA3: 'SUITEKA3',
    R_YCRU: 'YCRU',
    R_I18N: 'I18N',
    R_USTTP: 'US-TTP',
    R_CNTOB: 'CN-TOB',
    R_USEast: 'US-East',
    R_USBOE: 'US-BOE',
    R_SingaporeCentral: 'Singapore-Central',
    R_AsiaSouth: 'Asia-South',
    R_AsiaSaaS: 'Asia-SaaS',
    R_EuropeCentral: 'Europe-Central',
    R_QMSG: 'QMSG',
    R_USTTP2: 'US-TTP2',
    R_USCentral: 'US-Central',
    R_EUTTP: 'EU-TTP',
    R_ChinaPay: 'China-Pay',
    R_ChinaVolcEast: 'ChinaVolc-East',
    R_ChinaVolcSouth: 'ChinaVolc-South',
    R_ChinaEast: 'China-East',
    R_EUCOMPLIANCE2: 'EU-Compliance2',
    R_EUCOMPLIANCE: 'EU-Compliance',
    R_IDCompliance: 'ID-Compliance',
    R_IDCompliance2: 'ID-Compliance2',
    R_USCompliance: 'US-Compliance',
    R_EUTTP2: 'EU-TTP2',
    R_ChinaPay2: 'China-Pay2',
    R_AsiaSouthEastBD: 'Asia-SouthEastBD',
    R_USEastBD: 'US-EastBD',
    R_IDCOMPLIANCE2: 'ID-Compliance2',
    R_IDCOMPLIANCE: 'ID-Compliance',
    R_USCOMPLIANCE: 'US-Compliance',
    R_AsiaCIS: 'Asia-CIS',
    R_MYCompliance: 'MY-Compliance',
};
const dcToRegionMap = {
    [idc_util_1.IDC.DC_HY]: exports.REGION.R_CN,
    [idc_util_1.IDC.DC_LF]: exports.REGION.R_CN,
    [idc_util_1.IDC.DC_HL]: exports.REGION.R_CN,
    [idc_util_1.IDC.DC_YG]: exports.REGION.R_CN,
    [idc_util_1.IDC.DC_WJ]: exports.REGION.R_CN,
    [idc_util_1.IDC.DC_SDQD]: exports.REGION.R_CN,
    [idc_util_1.IDC.DC_SDQDCJ]: exports.REGION.R_CN,
    [idc_util_1.IDC.DC_ALINC2]: exports.REGION.R_CN,
    [idc_util_1.IDC.DC_LQ]: exports.REGION.R_CN,
    [idc_util_1.IDC.DC_AGSY]: exports.REGION.R_CN,
    [idc_util_1.IDC.DC_AGCQ]: exports.REGION.R_CN,
    [idc_util_1.IDC.DC_AGGDSZ]: exports.REGION.R_CN,
    [idc_util_1.IDC.DC_AGHBWH]: exports.REGION.R_CN,
    [idc_util_1.IDC.DC_AGJSNJ]: exports.REGION.R_CN,
    [idc_util_1.IDC.DC_AGLNSY]: exports.REGION.R_CN,
    [idc_util_1.IDC.DC_AGSDQD]: exports.REGION.R_CN,
    [idc_util_1.IDC.DC_AGSXXA]: exports.REGION.R_CN,
    [idc_util_1.IDC.DC_AGSXLQ]: exports.REGION.R_CN,
    [idc_util_1.IDC.DC_CA]: exports.REGION.R_CA,
    [idc_util_1.IDC.DC_MALIVA]: exports.REGION.R_MALIVA,
    [idc_util_1.IDC.DC_USEAST3]: exports.REGION.R_MALIVA,
    [idc_util_1.IDC.DC_USEAST4]: exports.REGION.R_MALIVA,
    [idc_util_1.IDC.DC_USEAST10A]: exports.REGION.R_MALIVA,
    [idc_util_1.IDC.DC_USEAST2A]: exports.REGION.R_I18N,
    [idc_util_1.IDC.DC_USEAST2B]: exports.REGION.R_I18N,
    [idc_util_1.IDC.DC_SG]: exports.REGION.R_SG,
    [idc_util_1.IDC.DC_SGEA1]: exports.REGION.R_QMSG,
    [idc_util_1.IDC.DC_SGEA2]: exports.REGION.R_QMSG,
    [idc_util_1.IDC.DC_SGEE1]: exports.REGION.R_QMSG,
    [idc_util_1.IDC.DC_SGEE2]: exports.REGION.R_QMSG,
    [idc_util_1.IDC.DC_SGSAAS1LARKIDC1]: exports.REGION.R_QMSG,
    [idc_util_1.IDC.DC_SGSAAS1LARKIDC2]: exports.REGION.R_QMSG,
    [idc_util_1.IDC.DC_BOE]: exports.REGION.R_BOE,
    [idc_util_1.IDC.DC_IBOE]: exports.REGION.R_BOE,
    [idc_util_1.IDC.DC_COF]: exports.REGION.R_BOE,
    [idc_util_1.IDC.DC_BOE2]: exports.REGION.R_ChinaBOE2,
    [idc_util_1.IDC.DC_BOETTP]: exports.REGION.R_USBOE,
    [idc_util_1.IDC.DC_VA]: exports.REGION.R_US,
    [idc_util_1.IDC.DC_ALISG]: exports.REGION.R_ALISG,
    [idc_util_1.IDC.DC_SG1]: exports.REGION.R_ALISG,
    [idc_util_1.IDC.DC_SGCOMM1]: exports.REGION.R_ALISG,
    [idc_util_1.IDC.DC_SGCOMPLIANCE]: exports.REGION.R_ALISG,
    [idc_util_1.IDC.DC_MY]: exports.REGION.R_ALISG,
    [idc_util_1.IDC.DC_MY2]: exports.REGION.R_ALISG,
    [idc_util_1.IDC.DC_MY3]: exports.REGION.R_ALISG,
    [idc_util_1.IDC.DC_SYKALARK]: exports.REGION.R_SUITEKA,
    [idc_util_1.IDC.DC_SYKA2LARK]: exports.REGION.R_SUITEKA2,
    [idc_util_1.IDC.DC_SYKA3LARK]: exports.REGION.R_SUITEKA3,
    [idc_util_1.IDC.DC_YCRU]: exports.REGION.R_YCRU,
    [idc_util_1.IDC.DC_USEAST5]: exports.REGION.R_USTTP,
    [idc_util_1.IDC.DC_USEAST8]: exports.REGION.R_USTTP2,
    [idc_util_1.IDC.DC_USORDX]: exports.REGION.R_USCentral,
    [idc_util_1.IDC.DC_LFTOBIAAS]: exports.REGION.R_CNTOB,
    [idc_util_1.IDC.DC_HLTOBIAAS]: exports.REGION.R_CNTOB,
    [idc_util_1.IDC.DC_NTTOBIAAS]: exports.REGION.R_CNTOB,
    [idc_util_1.IDC.DC_GZTOBIAAS]: exports.REGION.R_CNTOB,
    [idc_util_1.IDC.DC_USEASTDT]: exports.REGION.R_USEast,
    [idc_util_1.IDC.DC_USTSENTRY]: exports.REGION.R_USEast,
    [idc_util_1.IDC.DC_SGDT]: exports.REGION.R_SingaporeCentral,
    [idc_util_1.IDC.DC_SG2]: exports.REGION.R_SingaporeCentral,
    [idc_util_1.IDC.DC_IND]: exports.REGION.R_AsiaSouth,
    [idc_util_1.IDC.DC_INDDP]: exports.REGION.R_AsiaSouth,
    [idc_util_1.IDC.DC_JPSAAS]: exports.REGION.R_AsiaSaaS,
    [idc_util_1.IDC.DC_VOLCNT]: exports.REGION.R_ChinaVolcEast,
    [idc_util_1.IDC.DC_VOLCGZ]: exports.REGION.R_ChinaVolcSouth,
    [idc_util_1.IDC.DC_AWSFR]: exports.REGION.R_EuropeCentral,
    [idc_util_1.IDC.DC_FRACOMPLIANCE]: exports.REGION.R_EuropeCentral,
    [idc_util_1.IDC.DC_IE]: exports.REGION.R_EUTTP,
    [idc_util_1.IDC.DC_IEDT]: exports.REGION.R_EUTTP,
    [idc_util_1.IDC.DC_DEDT]: exports.REGION.R_EUTTP,
    [idc_util_1.IDC.DC_NO1A]: exports.REGION.R_EUTTP2,
    [idc_util_1.IDC.DC_IE2]: exports.REGION.R_EUCOMPLIANCE2,
    [idc_util_1.IDC.DC_DE]: exports.REGION.R_EUCOMPLIANCE,
    [idc_util_1.IDC.DC_ID1A]: exports.REGION.R_IDCompliance,
    [idc_util_1.IDC.DC_ID2A]: exports.REGION.R_IDCompliance2,
    [idc_util_1.IDC.DC_USEAST11A]: exports.REGION.R_USCompliance,
    [idc_util_1.IDC.DC_LFZG]: exports.REGION.R_ChinaPay,
    [idc_util_1.IDC.DC_HLZG]: exports.REGION.R_ChinaPay,
    [idc_util_1.IDC.DC_HJZG]: exports.REGION.R_ChinaPay2,
    [idc_util_1.IDC.DC_PD]: exports.REGION.R_ChinaEast,
    [idc_util_1.IDC.DC_HJ]: exports.REGION.R_ChinaEast,
    [idc_util_1.IDC.DC_MYA]: exports.REGION.R_AsiaSouthEastBD,
    [idc_util_1.IDC.DC_MYB]: exports.REGION.R_AsiaSouthEastBD,
    [idc_util_1.IDC.DC_USEAST9A]: exports.REGION.R_USEastBD,
    [idc_util_1.IDC.DC_MYCISB]: exports.REGION.R_AsiaCIS,
    [idc_util_1.IDC.DC_SGCISA]: exports.REGION.R_AsiaCIS,
    [idc_util_1.IDC.DC_MY4A]: exports.REGION.R_MYCompliance,
};
const regionFileDefault = '/opt/tiger/chadc/region.json';
const regionRefreshDur = 5 * 60 * 1000; // 5min
let region = ''; // local region (string)
let regionFile = ''; // file name (string)
let idcRegion = {}; // key is IDC, value is region (map[string]string)
let regionRefreshIntervalId = null;
const init = () => {
    region = '';
    regionFile = regionFileDefault;
    idcRegion = {};
    if (!regionRefreshIntervalId) {
        refreshRegion();
    }
    refreshRegionSync();
};
exports.init = init;
/**
 * Region .
 */
const getRegion = () => {
    if (region) {
        return region;
    }
    const { getIDC } = require('./idc');
    const idc = getIDC();
    const regionResult = (0, exports.getRegionFromIDC)(idc);
    region = regionResult;
    return regionResult;
};
exports.getRegion = getRegion;
/**
 * GetRegionFromIDC .
 */
const getRegionFromIDC = (idc) => {
    if (dcToRegionMap[idc]) {
        return dcToRegionMap[idc];
    }
    if (dcToRegionMap[idc_util_1.IDC[idc]]) {
        return dcToRegionMap[idc_util_1.IDC[idc]];
    }
    if (idcRegion[idc]) {
        return idcRegion[idc];
    }
    return exports.REGION.UnknownRegion;
};
exports.getRegionFromIDC = getRegionFromIDC;
/**
 * SetRegionFile .
 */
const setRegionFile = (file) => {
    regionFile = file;
    updateRegionIDCsSync();
};
exports.setRegionFile = setRegionFile;
/**
 * hasIDC return true if idc is support
 */
const hasIDC = (idc) => {
    return Boolean(idcRegion[idc]);
};
exports.hasIDC = hasIDC;
/**
 * idcList return IDCs which supported
 */
const idcList = () => {
    return Object.keys(idcRegion);
};
exports.idcList = idcList;
/**
 * GetIDCList .
 */
const getIDCList = () => {
    return (0, exports.idcList)();
};
exports.getIDCList = getIDCList;
/**
 * ATTENTION: IT COMES WITH A LOOP, DON'T CALL IT AGAIN.
 */
const refreshRegion = () => {
    regionRefreshIntervalId = setTimeout(refreshRegion, regionRefreshDur).unref();
    updateRegionIDCs().catch(noop);
};
const updateRegionIDCs = async () => {
    const file = regionFile;
    try {
        const content = await readFileAsync(file, { encoding: 'utf-8' });
        // newRegionIDCs, key is region, value is []IDC
        const newRegionIDCs = JSON.parse(content);
        const newIDCRegion = {};
        const keys = Object.keys(newRegionIDCs);
        for (let i = 0; i < keys.length; i++) {
            const region = keys[i];
            const dcs = newRegionIDCs[region];
            for (let i = 0; i < dcs.length; i++) {
                newIDCRegion[dcs[i]] = region;
            }
        }
        idcRegion = newIDCRegion;
    }
    catch (err) {
        // do nothing
    }
};
// 同步刷新，保证 env 可用
const refreshRegionSync = () => {
    updateRegionIDCsSync();
};
const updateRegionIDCsSync = () => {
    const file = regionFile;
    try {
        const content = readFileSync(file, { encoding: 'utf-8' });
        // newRegionIDCs, key is region, value is []IDC
        const newRegionIDCs = JSON.parse(content);
        const newIDCRegion = {};
        const keys = Object.keys(newRegionIDCs);
        for (let i = 0; i < keys.length; i++) {
            const region = keys[i];
            const dcs = newRegionIDCs[region];
            for (let i = 0; i < dcs.length; i++) {
                newIDCRegion[dcs[i]] = region;
            }
        }
        idcRegion = newIDCRegion;
    }
    catch (err) {
        // do nothing
    }
};
