// @universe-design/react/src/grid/col.tsx
import React, { useContext } from 'react';
import clsx from 'clsx';
import RowContext from './context';
import { useCssScope } from '../config-provider/hooks';
import type { Breakpoint } from '..';

import './style';

/**
 * 栅格数
 * @public
 */
export type ColSpanType = number | string;
export interface ColSize {
  span?: ColSpanType;
  order?: ColSpanType;
  offset?: ColSpanType;
  push?: ColSpanType;
  pull?: ColSpanType;
}

/**
 * @public
 */
export type MediaType = ColSpanType | ColSize;

/**
 * @public
 */
export type FlexType = number | 'none' | 'auto' | string;
/**
 * @public
 */
export interface ColProps extends React.HTMLAttributes<HTMLDivElement>, React.PropsWithChildren<unknown> {
  /**
   * 栅格占位格数，为 0 时相当于`display: none`
   */
  span?: ColSpanType;

  /**
   * 栅格顺序
   */
  order?: ColSpanType;
  /**
   * 栅格左侧的间隔格数，间隔内不可以有栅格
   */
  offset?: ColSpanType;
  /**
   * 栅格向右移动格数
   */
  push?: ColSpanType;
  /**
   * 栅格向左移动格数
   */
  pull?: ColSpanType;
  /**
   * 屏幕 < 576px响应式栅格，可为栅格数或一个包含其他属性的对象
   */
  xs?: MediaType;
  /**
   * 屏幕 ≥ 576px响应式栅格，可为栅格数或一个包含其他属性的对象
   */
  sm?: MediaType;
  /**
   * 屏幕 ≥ 768px响应式栅格，可为栅格数或一个包含其他属性的对象
   */
  md?: MediaType;
  /**
   * 屏幕 ≥ 992px响应式栅格，可为栅格数或一个包含其他属性的对象
   */
  lg?: MediaType;
  /**
   * 屏幕 ≥ 1200px响应式栅格，可为栅格数或一个包含其他属性的对象
   */
  xl?: MediaType;
  /**
   * 屏幕 ≥ 1600px响应式栅格，可为栅格数或一个包含其他属性的对象
   */
  xxl?: MediaType;

  /**
   * flex 布局属性
   */
  flex?: FlexType;
}

const breakpoints: Breakpoint[] = ['xs', 'sm', 'md', 'lg', 'xl', 'xxl'];

function parseFlex(flex: FlexType): string {
  if (typeof flex === 'number') {
    return `${flex} ${flex} auto`;
  }

  if (/^\d+(\.\d+)?(px|em|rem|%)$/.test(flex)) {
    return `0 0 ${flex}`;
  }

  return flex;
}

export const Col = React.forwardRef<HTMLDivElement, ColProps>((props, ref) => {
  const { span, order, offset, push, pull, children, xs, sm, md, lg, xl, xxl, flex, style, className, ...otherProps } =
    props;

  const cssScope = useCssScope('col');

  const { gutter, wrap } = useContext(RowContext);

  const rowStyle = React.useMemo(() => {
    const result: React.CSSProperties = {};
    if (gutter) {
      const [left, top] = gutter;
      if (left) {
        result.paddingLeft = left / 2;
        result.paddingRight = left / 2;
      }

      if (top) {
        result.paddingTop = top / 2;
        result.paddingBottom = top / 2;
      }
    }
    return result;
  }, [gutter]);

  const getMediaClass = () => {
    const points = { xs, sm, md, lg, xl, xxl };
    let result = {};
    breakpoints.forEach(p => {
      const point = points[p];
      if (typeof point === 'number') {
        result = {
          ...result,
          [cssScope(`-${p}-span-${point}`)]: point,
        };
      } else if (typeof point === 'object') {
        result = {
          ...result,
          [cssScope(`-${p}-span-${point.span}`)]: point.span !== undefined,
          [cssScope(`-${p}-order-${point.order}`)]: point.order || point.order === 0,
          [cssScope(`-${p}-offset-${point.offset}`)]: point.offset || point.offset === 0,
          [cssScope(`-${p}-push-${point.push}`)]: point.push || point.push === 0,
          [cssScope(`-${p}-pull-${point.pull}`)]: point.pull || point.pull === 0,
        };
      }
    });
    return result;
  };

  const mediacls = getMediaClass();

  const cls = clsx(
    cssScope(),
    {
      [cssScope(`-span-${span}`)]: span !== undefined,
      [cssScope(`-order-${order}`)]: order,
      [cssScope(`-offset-${offset}`)]: offset,
      [cssScope(`-push-${push}`)]: push,
      [cssScope(`-pull-${pull}`)]: pull,
    },
    mediacls,
    className,
  );

  const mergedStyle: React.CSSProperties = { ...rowStyle, ...style };

  if (flex) {
    mergedStyle.flex = parseFlex(flex);
    if (flex === 'auto' && wrap === false && !mergedStyle.minWidth) {
      mergedStyle.minWidth = 0;
    }
  }

  return (
    <div className={cls} style={mergedStyle} ref={ref} {...otherProps}>
      {children}
    </div>
  );
});
if (__DEV__) {
  Col.displayName = 'Col';
}
