// @universe-design/react/src/grid/row.tsx
import * as React from 'react';
import clsx from 'clsx';
import RowContext from './context';
import { useCssScope } from '../config-provider/hooks';
import type { Breakpoint, ScreenMap } from '../util';
import { responsiveArray, ResponsiveObserve } from '../util';
import type { Gutter, AlignType, JustifyType } from './interface';

import './style';
/**
 * @public
 */
export interface RowProps extends React.HTMLAttributes<HTMLDivElement>, React.PropsWithChildren<unknown> {
  /**
   * 栅格间隔，可以写成像素值或支持响应式的对象写法来设置水平间隔
   * { xs: 8, sm: 16, md: 24}。或者使用数组形式同时设置[水平间距, 垂直间距]
   */
  gutter?: Gutter | [Gutter, Gutter];

  /**
   * 垂直对齐方式
   */
  align?: AlignType;

  /**
   * 水平排列方式
   */
  justify?: JustifyType;
  prefixCls?: string;

  /**
   * 是否自动换行
   */
  wrap?: boolean;
}

export const Row = React.forwardRef<HTMLDivElement, RowProps>((props, ref) => {
  const { children, align, justify, gutter = 0, wrap, style, className, ...otherProps } = props;

  const cssScope = useCssScope('row');

  const classes = clsx(
    cssScope(),
    {
      [cssScope(`--align-${align}`)]: align,
      [cssScope(`--justify-${justify}`)]: justify,
      [cssScope('--no-wrap')]: wrap === false,
    },
    className,
  );

  const [screens, setScreens] = React.useState<ScreenMap>({
    xs: true,
    sm: true,
    md: true,
    lg: true,
    xl: true,
    xxl: true,
  });

  const gutterRef = React.useRef<Gutter | [Gutter, Gutter]>(gutter);

  React.useEffect(() => {
    const token = ResponsiveObserve.subscribe(screen => {
      const currentGutter = gutterRef.current || 0;
      if (
        (!Array.isArray(currentGutter) && typeof currentGutter === 'object') ||
        (Array.isArray(currentGutter) && (typeof currentGutter[0] === 'object' || typeof currentGutter[1] === 'object'))
      ) {
        setScreens(screen);
      }
    });
    return () => ResponsiveObserve.unsubscribe(token);
  }, []);

  const getGutter = (): [number, number] => {
    const results: [number, number] = [0, 0];
    const normalizedGutter = Array.isArray(gutter) ? gutter : [gutter, 0];
    normalizedGutter.forEach((g, index) => {
      if (typeof g === 'object') {
        for (let i = 0; i < responsiveArray.length; i++) {
          const breakpoint: Breakpoint = responsiveArray[i];
          if (screens[breakpoint] && g[breakpoint] !== undefined) {
            results[index] = g[breakpoint] as number;
            break;
          }
        }
      } else {
        results[index] = g || 0;
      }
    });
    return results;
  };

  const gutters = getGutter();

  const rowStyle = React.useMemo(() => {
    const result: React.CSSProperties = {};
    const [left, top] = gutters;
    if (left) {
      result.marginLeft = -left / 2;
      result.marginRight = -left / 2;
    }

    if (top) {
      result.marginTop = -top / 2;
      result.marginBottom = top / 2;
    }
    return result;
  }, [gutters]);

  return (
    <RowContext.Provider value={{ gutter: gutters, wrap }}>
      <div ref={ref} style={{ ...rowStyle, ...style }} className={classes} {...otherProps}>
        {children}
      </div>
    </RowContext.Provider>
  );
});
if (__DEV__) {
  Row.displayName = 'Row';
}

export type { AlignType, JustifyType, Gutter } from './interface';
