// @universe-design/react/es/expandable-text/ExpandableText.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.u10pi13n{position:relative;}
import u5jxv71 from "./../.stylex/u5jxv71";
//.u1xynx9j{left:0px;}
import u184z4yl from "./../.stylex/u184z4yl";
//.u10k790i{right:0px;}
import u5j3c7v from "./../.stylex/u5j3c7v";
//.u1euv43f{position:absolute;}
import u1u1p7mj from "./../.stylex/u1u1p7mj";
//.u5gq2j6{bottom:0px;}
import u8rtfz1 from "./../.stylex/u8rtfz1";
//.u9c4gz4{justify-content:flex-end;}
import uktrixa from "./../.stylex/uktrixa";
//.ubhxue7{justify-content:flex-start;}
import u1w892lm from "./../.stylex/u1w892lm";
//.u22iagw{display:flex;}
import ukchsq7 from "./../.stylex/ukchsq7";
//.ujw5fx7{width:16px;}
import ut0j16j from "./../.stylex/ut0j16j";
//.uebqbo6{-webkit-mask-image:linear-gradient(90deg, rgba(0, 0, 0, 0) 0%, rgba(0, 0, 0, 1) 100%);}
import u19pelrb from "./../.stylex/u19pelrb";
//.u11fbdcc{background-color:@bg-body;}
import u1to0gca from "./../.stylex/u1to0gca";
//.u1sil6mw{overflow-y:hidden;}
import u124hjsl from "./../.stylex/u124hjsl";
//.u1lmfglv{margin-top:0px;}
import u1kxgv9m from "./../.stylex/u1kxgv9m";
//.uibjyge{margin-right:0px;}
import unbze3l from "./../.stylex/unbze3l";
//.u1abmfm4{margin-bottom:0px;}
import u10d06an from "./../.stylex/u10d06an";
//.u9yszdx{margin-left:0px;}
import u1w6ojc6 from "./../.stylex/u1w6ojc6";
/** @jsxRuntime classic */
/** @jsx jsx */
import stylex from '@universe-design/stylex';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx, useMergeComponentProps } from '@universe-design/stylex/react';
import * as React from 'react';
import clamp from 'lodash/clamp';
import DownOutlined from '@universe-design/icons-react/DownOutlined';
import { useImmer } from 'use-immer';
import { useCssScope, useDirection, useLocaleReceiver } from '../config-provider';
import { forwardHTMLRef } from '../util/forwardRef';
import { Button } from '../button';
import { EmptyElement } from '../shared/components/FirstLineAlign';
import { useIsomorphicLayoutEffect as useLayoutEffect } from '../hooks';
import { useResizeObserver } from '@universe-design/react-utilities';
import { ExpandableProvider, useExpandableContext } from './Expandable';
import './style/text-collapse.less';
const useExpandableTextCssScope = () => useCssScope('expandable-text');

/**
 * @public
 * @docgen
 */

const ExpandableTextParagraph = props => {
  var _lineBottoms$lineClam;
  const {
    lineClamp: lineClampProp = 3,
    children
  } = props;
  const lineClamp = clamp(lineClampProp, 1, Infinity);
  const {
    isExpanded,
    updateOverflowed
  } = useExpandableContext();
  const cssScope = useExpandableTextCssScope();
  const paragraphRef = useResizeObserver(() => {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    reCalcLineBottoms();
  });
  const paragraphContentRef = React.useRef(null);

  /**
   * State
   */
  const [{
    lineBottoms
  }, updateState] = useImmer({
    lineBottoms: [0]
  });
  const reCalcLineBottoms = React.useCallback(() => {
    const paragraph = paragraphRef.current;
    const paragraphContent = paragraphContentRef.current;
    if (!paragraph || !paragraphContent) {
      return;
    }
    // 获取真实占位的 boundary
    const {
      top,
      bottom,
      height
    } = paragraph.getBoundingClientRect();
    const scaleY = height / paragraph.offsetHeight;
    const lineRects = [...paragraphContent.getClientRects()].filter(rect => rect.height && rect.width // firefox 在子元素是 div 时会多一个 width 是 0 的 rect，我们要排除掉
    );
    // console.trace('reCalcLineBottoms', { scaleY, top, bottom, lineRects });
    updateState(draft => {
      let prevLine = {
        index: 0,
        bottom: 0
      };
      draft.lineBottoms.splice(1 + lineRects.length);
      draft.lineBottoms[0] = 0;
      for (const [i, rect] of lineRects.entries()) {
        const spacingBottom = (() => {
          // eslint-disable-next-line max-nested-callbacks
          const nextLine = lineRects.slice(i + 1).find(r => r.top > rect.top); // 获取当前 rect 后第一个不同行的 rect，如果找到则把两行间隔的一半作为分界
          if (nextLine) {
            return (nextLine.top - rect.bottom) / 2;
          }
          return bottom - rect.bottom; // 如果没有找到不同行的 rect 说明是最后一行了，这时用 wrapper 的 bottom 作为边界
        })();
        /**
         * currentLineBottom 是相对于 top 的底部坐标，相当于 height，可以直接除 scaleY 得到 layout 尺寸
         */
        const currentLineBottom = Math.round((rect.bottom - top + spacingBottom) / scaleY); // 计算出当前行的下边界
        if (currentLineBottom === prevLine.bottom) {
          continue; // 如果和上次计算出的行下边界一致则说明这次的 rect 与之前的 rect 同行，这种情况出现在 white-space: pre-line 时，换行符 \n 会产生一个同行的零宽 rect
        }
        prevLine = {
          index: prevLine.index + 1,
          bottom: currentLineBottom
        };
        draft.lineBottoms[prevLine.index] = prevLine.bottom;
        // console.log('draft.lineBottoms', JSON.stringify(draft.lineBottoms));
      }
    });
    updateOverflowed(lineRects.length > lineClamp);
  }, [lineClamp, paragraphRef, updateOverflowed, updateState]);

  // lineClamp 变化的时候可以保证 reactive
  useLayoutEffect(() => {
    reCalcLineBottoms();
  }, [reCalcLineBottoms]);
  return jsx("div", {
    className: cssScope('__paragraph'),
    sx: [{
      "props": {
        "B6of3ja": u1kxgv9m,
        "t21cq0": unbze3l,
        "jrapky": u10d06an,
        "Frg6f3": u1w6ojc6
      }
    }, !isExpanded && {
      "props": {
        "Bmxbyg5": u124hjsl
      }
    }],
    style: {
      ...(isExpanded ? {} : {
        maxHeight: (_lineBottoms$lineClam = lineBottoms[lineClamp]) !== null && _lineBottoms$lineClam !== void 0 ? _lineBottoms$lineClam : lineBottoms[lineBottoms.length - 1] // 这里不直接用 webkit-line-clamp 实现是因为 safari 最后一行的底部没有 spacing 导致 trigger 定位会偏上
      })
    }
  }, jsx("div", {
    ref: paragraphRef
  }, jsx("span", {
    ref: paragraphContentRef,
    className: cssScope('__paragraph__content')
  }, children)));
};
const ExpandableTextTrigger = props => {
  const {
    maskColor,
    position: triggerPosition = 'end'
  } = props;
  const cssScope = useExpandableTextCssScope();
  const direction = useDirection();
  const {
    isExpanded,
    shouldHideTrigger,
    toggleExpanded
  } = useExpandableContext();
  const locale = useLocaleReceiver('expandableText');
  const shouldShowMask = triggerPosition === 'end' && !isExpanded;

  /**
   * Render
   */
  const maskNode = React.useMemo(() => {
    if (shouldShowMask) {
      return jsx("div", {
        className: cssScope('__trigger__mask'),
        sx: [{
          "props": {
            "mc9l5x": ukchsq7,
            "a9b677": ut0j16j,
            "Bx710cf": u19pelrb,
            "De3pzq": u1to0gca
          }
        }],
        style: {
          backgroundColor: maskColor
        }
      }, jsx(EmptyElement, null));
    }
    return null;
  }, [maskColor, cssScope, shouldShowMask]);
  const buttonNode = React.useMemo(() => jsx("div", {
    sx: [{
      "props": {
        "mc9l5x": ukchsq7
      }
    }, shouldShowMask && {
      "props": {
        "De3pzq": u1to0gca
      }
    }],
    style: {
      ...(shouldShowMask && {
        backgroundColor: maskColor
      })
    }
  }, jsx(Button, {
    className: cssScope('__trigger__button'),
    type: "link"
    // 这里用 `DownOutlined` 而不是 `DownBoldOutlined` 是因为这里细一点好看
    ,
    endIcon: jsx(DownOutlined, {
      style: {
        ...(isExpanded && {
          transform: 'rotate(180deg)'
        })
      }
    }),
    onClick: toggleExpanded
  }, isExpanded ? locale.fold : locale.show)), [shouldShowMask, maskColor, cssScope, isExpanded, toggleExpanded, locale.fold, locale.show]);
  if (shouldHideTrigger) {
    return null;
  }
  return jsx("div", {
    className: cssScope('__trigger'),
    sx: [{
      "props": {
        "mc9l5x": ukchsq7
      }
    }, triggerPosition === 'start' && [{
      "props": {
        "Brf1p80": u1w892lm
      }
    }], triggerPosition === 'end' && [isExpanded && [{
      "props": {
        "Brf1p80": uktrixa
      }
    }], !isExpanded && [{
      "props": {
        "qhf8xq": u1u1p7mj,
        "B5kzvoi": u8rtfz1
      }
    }, direction === 'ltr' ? {
      "props": {
        "j35jbq": u5j3c7v
      }
    } : {
      "props": {
        "oyh7mz": u184z4yl
      }
    }]]]
  }, maskNode, buttonNode);
};

/**
 * 文字收起组件
 */
export const ExpandableText = forwardHTMLRef((props, ref) => {
  const {
    children,
    preventCollapseWhenExpanded,
    lineClamp,
    triggerProps,
    ...domProps
  } = props;

  /**
   * Context
   */
  const cssScope = useExpandableTextCssScope();

  /**
   * Props
   */
  const mergedDOMProps = useMergeComponentProps(domProps, {
    sx: [{
      "props": {
        "qhf8xq": u5jxv71
      },
      "alias": cssScope()
    }]
  });
  return jsx(ExpandableProvider, {
    preventCollapseWhenExpanded
  }, jsx("div", _extends({
    ref: ref
  }, mergedDOMProps), jsx(ExpandableTextParagraph, {
    lineClamp: lineClamp
  }, children), jsx(ExpandableTextTrigger, triggerProps)));
});
if (process.env.NODE_ENV !== "production") {
  ExpandableText.displayName = 'ExpandableText';
}

/**
 * @public
 * @docgen
 */
