// @universe-design/react/es/expandable-text/Expandable.js
import * as React from 'react';
import { useImmer } from 'use-immer';
import { useEventCallback } from '../hooks';
import { createContextHook, useControlledState } from '@universe-design/react-utilities';
const useExpandable = props => {
  var _isExpandedProp$defau;
  const {
    isExpanded: isExpandedProp = {},
    preventCollapseWhenExpanded = false
  } = props;

  /**
   * State
   */
  const [{
    isOverflowed
  }, updateState] = useImmer({
    isOverflowed: false
  });

  /**
   * isExpanded State
   */
  const [isExpanded, setExpanded] = useControlledState((_isExpandedProp$defau = isExpandedProp.defaultValue) !== null && _isExpandedProp$defau !== void 0 ? _isExpandedProp$defau : false, isExpandedProp.value, {
    onChange: isExpandedProp.onChange
  });
  const toggleExpanded = useEventCallback(() => {
    setExpanded(!isExpanded);
  });
  const updateOverflowed = useEventCallback(_isOverflowed => {
    updateState(draft => {
      draft.isOverflowed = _isOverflowed;
    });
  });
  return {
    /**
     * 是否隐藏 trigger
     */
    shouldHideTrigger: isExpanded && preventCollapseWhenExpanded || !isOverflowed,
    isExpanded,
    toggleExpanded,
    updateOverflowed
  };
};
const {
  Provider: ExpandableContextProvider,
  useContext: useExpandableContext
} = createContextHook('ExpandableContext', {
  strict: true
});
export const ExpandableProvider = _ref => {
  let {
    children,
    ...props
  } = _ref;
  return /*#__PURE__*/React.createElement(ExpandableContextProvider, {
    value: useExpandable(props)
  }, children);
};
export { useExpandableContext };
