// @universe-design/react/lib/tree-select/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  TreeSelect: true
};
exports.TreeSelect = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _configProvider = require("../config-provider");
var _internal = require("../select/internal.interface");
var _select = require("../select");
var _common = require("./common");
var _tree = require("../tree");
var _util = require("../util");
var _common2 = require("../tree/common");
var _util2 = require("../tree/util");
var _TreeCheckedContext = require("../tree/hooks/TreeCheckedContext");
var _checkStrategy = require("../tree/hooks/checkStrategy");
var _useTreeCheck = require("../tree/hooks/useTreeCheck");
var _useTree = require("../tree/hooks/useTree");
var _TreeContainer = require("../tree/TreeContainer");
var _TreeContext = require("../tree/hooks/TreeContext");
var _SelectOverlayContainer = require("../select/SelectOverlayContainer");
require("./style");
var _interface = require("./interface");
Object.keys(_interface).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _interface[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _interface[key];
    }
  });
});
/* eslint-disable react-hooks/rules-of-hooks */
/* eslint-disable max-lines-per-function */

const RefSelect = (0, _select.selectGenerator)({
  selectorCssScope: 'select',
  dropdownCssScope: 'select',
  defaultOptionLabelProp: option => {
    var _ref, _option$label;
    return (_ref = (_option$label = option.label) !== null && _option$label !== void 0 ? _option$label : option.title) !== null && _ref !== void 0 ? _ref : option.value;
  }
});

/**
 *   因为 tree 和 select-tree的value参数不一样， 所以需要将 tree-select的value 格式化为tree的value
 *   string[] -> string[] | {checked: string[], unchecked: string[]}
 */

function toCheckStrictlyValue(value) {
  if (Array.isArray(value)) {
    return {
      checked: value,
      halfChecked: []
    };
  }
  return {
    checked: [],
    halfChecked: []
  };
}
const TreeSelectOverlay = _ref2 => {
  var _treeData$length;
  let {
    multiple,
    treeCheckable,
    treeData,
    treeDefaultExpandAll,
    treeDefaultExpandedKeys,
    treeExpandedKeys,
    onTreeExpand,
    listHeight,
    filterTreeNode: filterTreeNodeProp,
    showCheckedStrategy,
    treeCheckStrictly,
    switcherIcon,
    loadData,
    virtual,
    nodeRenderer,
    treeRef,
    treeLoadedKeys,
    onTreeLoad
  } = _ref2;
  const cssScope = (0, _configProvider.useCssScope)('treeSelect');
  const {
    inputValue,
    valueSet,
    set
  } = (0, _react.useContext)(_select.SelectContext);
  let treeProps = {
    ref: treeRef,
    searchValue: inputValue,
    defaultExpandAll: treeDefaultExpandAll,
    expandedKeys: treeExpandedKeys,
    defaultExpandedKeys: treeDefaultExpandedKeys,
    onExpand: onTreeExpand,
    multiple,
    selectable: !treeCheckable,
    checkable: treeCheckable,
    showChecker: !treeCheckable,
    treeData,
    selectedKeys: treeCheckable ? [] : [...valueSet],
    nodeRenderer,
    onSelect: (keys, _action) => {
      const action = {
        type: _action.selected ? _internal.SelectActionType.Select : _internal.SelectActionType.DeSelect,
        payload: _action.key
      };
      set(keys, {
        action
      });
    },
    // ==================== Filter Tree Node ====================
    filterTreeNode: (0, _react.useCallback)(node => {
      if (filterTreeNodeProp === false) {
        return true;
      } else if ((0, _util.isFunction)(filterTreeNodeProp)) {
        return filterTreeNodeProp(inputValue, node);
      }
      return (0, _common.defaultFilterTreeNode)(inputValue, node);
    }, [filterTreeNodeProp, inputValue]),
    showCheckedStrategy,
    switcherIcon,
    loadData,
    height: listHeight - 3 * 2 - 1 * 2,
    // 减去 padding border
    virtual,
    loadedKeys: treeLoadedKeys,
    onLoad: onTreeLoad
  };
  if (!treeCheckStrictly) {
    treeProps = {
      ...treeProps,
      checkedKeys: treeCheckable ? [...valueSet] : [],
      checkStrictly: false,
      onCheck: (keys, _action) => {
        const action = {
          type: _action.checked ? _internal.SelectActionType.Checked : _internal.SelectActionType.UnChecked,
          payload: _action.key
        };
        set(keys, {
          closeOverlay: false,
          action
        });
      }
    };
  } else {
    treeProps = {
      ...treeProps,
      defaultCheckedKeys: {
        checked: [],
        halfChecked: []
      },
      checkedKeys: treeCheckable ? {
        checked: [...valueSet],
        halfChecked: []
      } : {
        checked: [],
        halfChecked: []
      },
      checkStrictly: true,
      onCheck: (keys, _action) => {
        const action = {
          type: _action.checked ? _internal.SelectActionType.Checked : _internal.SelectActionType.UnChecked,
          payload: _action.key
        };
        set(keys.checked, {
          closeOverlay: false,
          action
        });
      }
    };
  }
  const treeCtxValue = (0, _useTree.useTree)(treeProps, treeRef);
  const menuNode = /*#__PURE__*/_react.default.createElement(_TreeContext.TreeContextProvider, {
    value: treeCtxValue
  }, /*#__PURE__*/_react.default.createElement(_TreeContainer.TreeContainer, null));
  return /*#__PURE__*/_react.default.createElement(_SelectOverlayContainer.SelectOverlayContainer, {
    className: cssScope('__overlay'),
    optionsLength: (_treeData$length = treeData === null || treeData === void 0 ? void 0 : treeData.length) !== null && _treeData$length !== void 0 ? _treeData$length : 0,
    filteredOptionsLength: treeCtxValue.filteredTreeNodes.size,
    menu: menuNode
  });
};

/**
 * @docgen
 */

const _TreeSelect = /*#__PURE__*/_react.default.forwardRef(function (props, ref) {
  var _props$multiple;
  const {
    treeData: treeDataProp,
    treeCheckable = false,
    treeDefaultExpandAll = false,
    treeDefaultExpandedKeys = [],
    treeExpandedKeys,
    onTreeExpand,
    listHeight = 232,
    filterTreeNode: filterTreeNodeProp = true,
    showCheckedStrategy = _tree.ShowCheckedStrategy.SHOW_CHILD,
    treeCheckStrictly = false,
    switcherIcon,
    loadData,
    children,
    virtual,
    nodeRenderer,
    treeLoadedKeys,
    onTreeLoad,
    ...baseSelectProps
  } = props;
  const selectRef = _react.default.useRef(null);
  const treeRef = _react.default.useRef(null);
  const treeData = (0, _react.useMemo)(() => treeDataProp !== null && treeDataProp !== void 0 ? treeDataProp : (0, _common2.treeChildren2TreeNodes)(children), [children, treeDataProp]);
  const {
    doublyTreeNodeMap
  } = (0, _react.useMemo)(() => (0, _util2.genDoublyTreeNodeMap)(treeData || []), [treeData]);
  const treeCheckCtx = (0, _useTreeCheck.useTreeCheck)({
    doublyTreeNodeMap,
    checkStrategy: _checkStrategy.strategyMap[showCheckedStrategy],
    ...(treeCheckStrictly ? {
      checkStrictly: true,
      defaultCheckedKeys: toCheckStrictlyValue(props.defaultValue),
      ...('value' in props ? {
        checkedKeys: toCheckStrictlyValue(props.value)
      } : {})
    } : {
      checkStrictly: false,
      defaultCheckedKeys: (0, _util.genArray)(props.defaultValue),
      ...('value' in props ? {
        checkedKeys: (0, _util.genArray)(props.value)
      } : {})
    })
  }, {
    skipped: !treeCheckable
  });
  const {
    checkedSet,
    setCheckedKeys
  } = treeCheckCtx;
  const checkedKeys = _react.default.useMemo(() => Object.keys(checkedSet).filter(key => checkedSet[key]), [checkedSet]);

  /**
   * Select Props
   */
  let multiple = (_props$multiple = props.multiple) !== null && _props$multiple !== void 0 ? _props$multiple : false;
  if (treeCheckable) {
    multiple = true;
  }
  baseSelectProps.multiple = multiple;
  const internalOptions = _react.default.useMemo(() => {
    const options = [];
    (0, _tree.traverseTreeConfig)(treeData, node => {
      options.push(node);
    });
    return options;
  }, [treeData]);
  const overlayNode = /*#__PURE__*/_react.default.createElement(_TreeCheckedContext.TreeCheckedProvider, {
    value: treeCheckCtx
  }, /*#__PURE__*/_react.default.createElement(TreeSelectOverlay, {
    treeRef: treeRef,
    treeData,
    treeCheckable,
    treeDefaultExpandAll,
    treeDefaultExpandedKeys,
    treeExpandedKeys,
    onTreeExpand,
    listHeight,
    multiple,
    filterTreeNode: filterTreeNodeProp,
    showCheckedStrategy,
    treeCheckStrictly,
    switcherIcon,
    loadData,
    virtual,
    nodeRenderer,
    treeLoadedKeys,
    onTreeLoad
  }));
  let selectProps;
  if (baseSelectProps.multiple) {
    if (treeCheckable) {
      if ('value' in props) {
        // 受控模式
        baseSelectProps.value = checkedKeys;
      } else {
        // 非受控模式
        baseSelectProps.defaultValue = checkedKeys;
      }
    }
    selectProps = {
      ...baseSelectProps,
      mode: 'multiple',
      options: internalOptions,
      overlayNode,
      onChange: (value, meta) => {
        if (treeCheckable) {
          const newValue = value;
          setCheckedKeys(treeCheckStrictly ? toCheckStrictlyValue(newValue) : (0, _util.genArray)(newValue));
        }
        if (baseSelectProps.onChange) {
          const path = value.length ? value.reduce((a, b, index) => {
            const _path = (0, _tree.getTreeNodeConfigPath)(doublyTreeNodeMap[b]);
            a[index] = _path;
            return a;
          }, []) : [];
          baseSelectProps.onChange(value, {
            ...meta,
            path
          });
        }
      }
    };
  } else {
    selectProps = {
      ...baseSelectProps,
      mode: 'single',
      options: internalOptions,
      overlayNode,
      onChange: (value, meta) => {
        if (baseSelectProps.onChange) {
          const path = (0, _tree.getTreeNodeConfigPath)((0, _util.isNonNullable)(value) ? doublyTreeNodeMap[value] : null);
          baseSelectProps.onChange(value, {
            ...meta,
            path
          });
        }
      }
    };
  }
  _react.default.useImperativeHandle(ref, () => ({
    get input() {
      var _selectRef$current;
      return (_selectRef$current = selectRef.current) === null || _selectRef$current === void 0 ? void 0 : _selectRef$current.input;
    },
    blur() {
      var _selectRef$current2;
      (_selectRef$current2 = selectRef.current) === null || _selectRef$current2 === void 0 ? void 0 : _selectRef$current2.blur();
    },
    focus() {
      var _selectRef$current3;
      (_selectRef$current3 = selectRef.current) === null || _selectRef$current3 === void 0 ? void 0 : _selectRef$current3.focus();
    },
    scrollIntoView() {
      var _treeRef$current;
      for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }
      (_treeRef$current = treeRef.current) === null || _treeRef$current === void 0 ? void 0 : _treeRef$current.scrollIntoView(...args);
    },
    toggleExpanded(key) {
      var _treeRef$current2;
      (_treeRef$current2 = treeRef.current) === null || _treeRef$current2 === void 0 ? void 0 : _treeRef$current2.toggleExpanded(key);
    }
  }));
  return /*#__PURE__*/_react.default.createElement(RefSelect, (0, _extends2.default)({
    ref: selectRef
  }, selectProps));
});
const TreeSelect = exports.TreeSelect = Object.assign(_TreeSelect, {
  SHOW_ALL: _tree.ShowCheckedStrategy.SHOW_ALL,
  SHOW_PARENT: _tree.ShowCheckedStrategy.SHOW_PARENT,
  SHOW_CHILD: _tree.ShowCheckedStrategy.SHOW_CHILD,
  TreeNode: _tree.TreeNode,
  [_configProvider.COMPONENT_VERSION_ID]: "3.28.3",
  displayName: 'TreeSelect'
});
