// @universe-design/react/src/upload/hooks/useUploadTrigger.ts
import React from 'react';
import uid from '../utils/uid';
import { file2Obj, getFileItem, LIST_IGNORE, updateFileList } from '../utils';
import defaultRequest from '../utils/request';
import { useEventCallback } from '../../hooks';
import type { UploadFile } from '../uploadFile/interface';
import type {
  ComposableUploadProps,
  RcFile,
  BeforeUploadFileType,
  UploadProgressEvent,
  UploadRequestError,
  UploadRequestOption,
} from '../interface';
interface ParsedFileInfo {
  origin: RcFile;
  action: string;
  data: Record<string, unknown> | null;
  parsedFile: RcFile | null;
}

export const useUploadTrigger = (
  props: ComposableUploadProps,
  ctxValue: {
    onChange: (
      file: UploadFile,
      changedFileList: UploadFile[],
      event?:
        | {
            percent: number;
          }
        | undefined,
    ) => void;
    fileListRef: React.MutableRefObject<UploadFile[]>;
  },
) => {
  const {
    headers,
    // multiple = false,
    withCredentials,
    customRequest,
    method,
    data,
    action,
    beforeUpload: beforeUploadProp,
    maxCount,
    onExceed,
    name = 'file',
  } = props;
  const { fileListRef, onChange } = ctxValue;
  type XhrReturn = { abort: () => void };
  const xhrRequests = React.useRef<Map<string, XhrReturn>>(new Map());
  const _isMounted = React.useRef(false);
  React.useEffect(() => {
    _isMounted.current = true;
    return () => {
      // this.abort();
      _isMounted.current = false;
    };
  }, []);

  const onError = useEventCallback((error: Error, response: any, file: RcFile) => {
    // removed
    if (!getFileItem(file, fileListRef.current)) {
      return;
    }

    const targetItem = file2Obj(file);
    targetItem.error = error;
    targetItem.response = response;
    targetItem.status = 'error';
    onChange(targetItem, updateFileList(targetItem, fileListRef.current));
  });

  const onProgress = useEventCallback((e: { percent: number; loaded: number }, file: RcFile) => {
    const fileItem = getFileItem(file, fileListRef.current);
    if (!fileItem) {
      return;
    }

    const targetItem = file2Obj(file);
    targetItem.status = 'uploading';
    targetItem.percent = e.percent;
    targetItem.loaded = e.loaded;

    const nextFileList = updateFileList(Object.assign(fileItem, targetItem), fileListRef.current);
    onChange(targetItem, nextFileList, e);
  });

  const beforeUpload = React.useCallback(
    async (file: RcFile, fileListArgs: RcFile[]) => {
      let parsedFile: File | Blob | string = file;
      if (beforeUploadProp) {
        const result = await beforeUploadProp(file, fileListArgs);
        if (result === false) {
          return false;
        }
        // Hack for LIST_IGNORE, we add additional info to remove from the list
        delete (file as any)[LIST_IGNORE];
        if ((result as any) === LIST_IGNORE) {
          Object.defineProperty(file, LIST_IGNORE, {
            value: true,
            configurable: true,
          });
          return false;
        }

        if (typeof result === 'object' && result) {
          parsedFile = result as File;
        }
      }
      return parsedFile;
    },
    [beforeUploadProp],
  );
  /**
   * Process file before upload. When all the file is ready, we start upload.
   */
  const processFile = React.useCallback(
    async (file: RcFile, _fileList: RcFile[]): Promise<ParsedFileInfo> => {
      let transformedFile: BeforeUploadFileType | void = file;
      if (beforeUpload && typeof beforeUpload === 'function') {
        try {
          transformedFile = await beforeUpload(file, _fileList);
        } catch (e) {
          // Rejection will also trade as false
          transformedFile = false;
        }
        if (!transformedFile) {
          return {
            origin: file,
            parsedFile: null,
            action: '',
            data: null,
          };
        }
      }

      // Get the latest action
      const mergedAction = typeof action === 'function' ? await action(file) : action || '';

      // Get latest data
      const mergedData = typeof data === 'function' ? await data(file) : data;

      const parsedData =
        // string type is from legacy `transformFile`.
        // Not sure if this will work since no related test case works with it
        (typeof transformedFile === 'object' || typeof transformedFile === 'string') && transformedFile
          ? transformedFile
          : file;

      const parsedFile =
        parsedData instanceof File ? parsedData : new File([parsedData], file.name, { type: file.type });

      const mergedParsedFile: RcFile = parsedFile as RcFile;
      mergedParsedFile.uid = file.uid;

      return {
        origin: file,
        data: mergedData,
        parsedFile: mergedParsedFile,
        action: mergedAction,
      };
    },
    [action, beforeUpload, data],
  );

  /**
   * 注意：
   *  onBatchStart onSuccess onError 会在上传的过程中异步触发，
   *  此处 用 useEventCallback 保证拿到最新值
   */
  const onBatchStart = useEventCallback(
    (
      batchFileInfoList: {
        file: RcFile;
        parsedFile: Exclude<BeforeUploadFileType, boolean>;
      }[],
    ) => {
      // Skip file which marked as `LIST_IGNORE`, these file will not add to file list
      const filteredFileInfoList = batchFileInfoList.filter(info => !(info.file as any)[LIST_IGNORE]);

      // Nothing to do since no file need upload
      if (!filteredFileInfoList.length) {
        return;
      }

      const objectFileList: UploadFile[] = filteredFileInfoList.map(info => ({
        ...file2Obj(info.file),
        status: 'waiting',
      }));

      // Concat new files with prev files
      let newFileList = [...fileListRef.current];

      objectFileList.forEach(fileObj => {
        // Replace file if exist
        newFileList = updateFileList(fileObj, newFileList);
      });

      if (maxCount && newFileList.length > maxCount) {
        onExceed?.();
      }

      objectFileList.forEach((fileObj, index) => {
        // Repeat trigger `onChange` event for compatible
        let triggerFileObj: UploadFile = fileObj;

        if (!filteredFileInfoList[index].parsedFile) {
          // `beforeUpload` return false
          const { originFileObj } = fileObj as { originFileObj: File };
          let clone;

          try {
            clone = new File([originFileObj as File], originFileObj.name, {
              type: originFileObj.type,
            }) as any as UploadFile;
          } catch (e) {
            clone = new Blob([originFileObj], {
              type: originFileObj.type,
            }) as any as UploadFile;
            clone.name = originFileObj.name;
            clone.lastModifiedDate = new Date();
            clone.lastModified = new Date().getTime();
          }

          clone.uid = fileObj.uid;
          triggerFileObj = clone;
        } else {
          // Inject `waiting` status
          fileObj.status = 'waiting';
        }

        onChange(triggerFileObj, newFileList);
      });
    },
  );

  const onSuccess = useEventCallback((response: any, file: RcFile, xhr: any) => {
    // removed
    if (!getFileItem(file, fileListRef.current)) {
      return;
    }

    try {
      if (typeof response === 'string') {
        response = JSON.parse(response);
      }
    } catch (e) {
      /* do nothing */
    }

    const targetItem = file2Obj(file);
    targetItem.status = 'success';
    targetItem.percent = 100;
    targetItem.response = response;
    targetItem.xhr = xhr;

    onChange(targetItem, updateFileList(targetItem, fileListRef.current));
  });

  const post = React.useCallback(
    ({ origin, parsedFile, data: _data, action: _action }: ParsedFileInfo) => {
      if (!_isMounted) {
        return;
      }

      const { uid: _uid } = origin;
      const request = customRequest || defaultRequest;

      const requestOption = {
        action: _action,
        filename: name,
        data: _data,
        file: parsedFile,
        headers,
        withCredentials,
        method: method || 'post',
        onProgress: (e: UploadProgressEvent) => {
          onProgress(e, parsedFile as RcFile);
        },
        onSuccess: (ret: any, xhr: XMLHttpRequest) => {
          onSuccess(ret, parsedFile as RcFile, xhr);
          xhrRequests.current.delete(_uid);
        },
        onError: (err: UploadRequestError, ret: any) => {
          onError(err, ret, parsedFile as RcFile);
          xhrRequests.current.delete(_uid);
        },
      };

      // onStart?.(origin);
      const aborter = request(requestOption as UploadRequestOption);
      if (aborter) {
        xhrRequests.current.set(_uid, aborter);
      }
    },
    [customRequest, headers, method, name, onError, onProgress, onSuccess, withCredentials],
  );

  const uploadFiles = React.useCallback(
    (files: File[]) => {
      const originFiles = [...files] as RcFile[];
      const postFiles = originFiles.map((file: RcFile & { uid?: string }) => {
        // 如果有uid 不改变uid
        file.uid = file.uid || uid();
        return processFile(file, originFiles);
      });

      // Batch upload files
      Promise.all(postFiles).then(_fileList => {
        onBatchStart(
          _fileList.map(({ origin, parsedFile }) => ({
            file: origin,
            parsedFile: parsedFile as RcFile,
          })),
        );

        _fileList
          .filter(file => file.parsedFile !== null)
          .forEach(file => {
            post(file);
          });
      });
    },
    [onBatchStart, post, processFile],
  );

  const abort = React.useCallback((file: any) => {
    if (file) {
      const _uid = file.uid ? file.uid : file;
      const reqCache = xhrRequests.current.get(_uid);
      if (reqCache && reqCache.abort) {
        reqCache.abort();
      }
      xhrRequests.current.delete(_uid);
    } else {
      Object.keys(xhrRequests).forEach(_uid => {
        const reqCache = xhrRequests.current.get(_uid);
        if (reqCache && reqCache.abort) {
          reqCache.abort();
        }
        xhrRequests.current.delete(_uid);
      });
    }
  }, []);

  const retry = React.useCallback(
    (file: RcFile) => {
      uploadFiles([file]);
    },
    [uploadFiles],
  );

  return {
    onBatchStart,
    onError,
    onProgress,
    onSuccess,
    uploadFiles,
    abort,
    retry,
  };
};
