// @byted-nodex/metrics/lib/client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Client = exports.defaultConfig = void 0;
const metric_1 = require("./metric");
const consts_1 = require("./consts");
const fs_1 = __importDefault(require("fs"));
const writer_1 = require("./writer");
const util_1 = require("./util");
const events_1 = require("events");
const logger_core_1 = require("@byted-nodex/logger-core");
const metrics_codec_1 = require("@byted-nodex/metrics-codec");
// 跟 v1 逻辑一致
const METRICS_ENABLE = process.env.METRICS_ENABLE !== 'false';
exports.defaultConfig = {
    defaultPrefix: 'toutiao',
    flushInterval: 30 * 1000, // 默认 30s
    maxRetryWriteTimes: 3, // 上报默认重试 3 次
    reportDefaultTags: true,
    metricGCOptions: {
        gcInterval: consts_1.defaultGCInterval,
        gcThreshold: consts_1.defaultGCThreshold,
        expireDuration: consts_1.defaultExpireDuration,
        forceGCCacheCursorCount: consts_1.defaultForceGCCursorCount,
        forceGCCacheMetricCount: consts_1.defaultForceGCCacheMetricCount,
        forceGCHealthCheckCount: consts_1.defaultForceGCHealthCheckCount,
    },
};
class Client extends events_1.EventEmitter {
    constructor(options) {
        super();
        this.inited = false;
        this.metrics = new Map();
        this.nextFlushTime = 0;
        this.forceGCCount = 0;
        this.closed = false;
        // 如果传了 logger 或者 debug 为 true，默认打印日志
        const silent = typeof (options === null || options === void 0 ? void 0 : options.silent) === 'boolean'
            ? options.silent
            : (options === null || options === void 0 ? void 0 : options.logger)
                ? false
                : !(options === null || options === void 0 ? void 0 : options.debug);
        const ops = (this.options = {
            silent,
            debug: false,
            ...exports.defaultConfig,
            ...options,
        });
        this.gcOptions = {
            ...exports.defaultConfig.metricGCOptions,
            ...options === null || options === void 0 ? void 0 : options.metricGCOptions,
        };
        // init logger
        this.logger = ops.silent ? util_1.silentLogger : ops.logger || new logger_core_1.ConsoleLogger();
        if (!(0, util_1.isValidString)(ops.defaultPrefix)) {
            throw new Error(`invalid prefix name: ${ops.defaultPrefix}`);
        }
        // init flush interval
        if (!ops.flushInterval || !consts_1.candidateTimeIntervals.includes(ops.flushInterval)) {
            this.logger.error(`flushInterval must be one of ${consts_1.candidateTimeIntervals} but receive ${ops.flushInterval}, reset to default config: ${exports.defaultConfig.flushInterval}`);
            ops.flushInterval = exports.defaultConfig.flushInterval;
        }
        // 单次上报请求次数
        this.requestTimes = this.options.maxRetryWriteTimes + 1;
        this.defaultTags = this.options.reportDefaultTags ? (0, util_1.getDefaultTags)() : {};
    }
    newMetricWithCache(metricName, options = {}) {
        options.withCache = true;
        return this.newMetric(metricName, options);
    }
    newMetric(metricName, options = {}) {
        if (!(0, util_1.isValidString)(metricName)) {
            throw new Error(`invalid metric name: ${metricName}`);
        }
        if (options.prefix && !(0, util_1.isValidString)(options.prefix)) {
            throw new Error(`invalid prefix name: ${options.prefix}`);
        }
        // lazy init
        this.init();
        const prefix = typeof options.prefix === 'string' ? options.prefix : this.options.defaultPrefix;
        const defaultMetricOptions = this.options.metricOptions;
        const metricOptions = {
            ...defaultMetricOptions,
            ...options,
            globalTags: {
                ...this.defaultTags,
                ...defaultMetricOptions === null || defaultMetricOptions === void 0 ? void 0 : defaultMetricOptions.globalTags,
                ...options.globalTags,
            },
            name: prefix ? `${prefix}.${metricName}` : metricName,
            client: this,
        };
        const metricId = options.withCache ? (0, util_1.serializeMetricOption)(metricOptions) : (0, util_1.genUniqId)();
        if (this.metrics.has(metricId)) {
            return this.metrics.get(metricId);
        }
        const metric = new metric_1.Metric(metricOptions);
        metric.id = metricId;
        this.metrics.set(metricId, metric);
        return metric;
    }
    /** 释放掉指定 metric */
    async releaseMetric(m) {
        if (!this.metrics.has(m.id))
            return;
        this.metrics.delete(m.id);
        await this.flushMetric(m).catch(util_1.noop);
        if (!this.metrics.size) {
            // 没有存量 metric ，将 client reset 掉
            await this.reset();
        }
    }
    init() {
        if (this.inited) {
            return;
        }
        // init writer pool
        const writerPoolOptions = this.options.writerPoolOptions || {};
        let writer = METRICS_ENABLE ? writerPoolOptions.writer : writer_1.NoopWriter;
        if (!writer) {
            const wps = (writerPoolOptions.writerOptions = writerPoolOptions.writerOptions || {});
            const socketPath = (wps.socketPath = wps.socketPath || (0, util_1.getSeqpacketSocketPath)());
            if (!fs_1.default.existsSync(socketPath)) {
                this.logger.warn(`Can not find seqpacket socket path: ${socketPath}, metrics writer is not available`);
                writer = writer_1.NoopWriter;
            }
            else {
                writer = writer_1.SeqpacketWriter;
            }
        }
        if (!METRICS_ENABLE) {
            this.logger.warn('Will NOT send metrics to server due to environment variables(METRICS_ENABLE=false).');
        }
        this.pool = new writer_1.WriterPool({
            logger: this.logger,
            writer,
            ...writerPoolOptions,
        });
        // start interval
        this.flushInterval();
        this.gcInterval();
        this.inited = true;
    }
    /** 开启 gc 的定时器 */
    gcInterval() {
        const gcInterval = Math.max(this.gcOptions.gcInterval, consts_1.minGCInterval);
        this.gcTicker = setInterval(() => this.gc(), gcInterval).unref();
    }
    /** 开启 flush 的定时器，需要根据 flushInterval 取整 flush，比如 11:30:10 -> 11:30:15 -> 11:30:20 */
    async flushInterval() {
        const [nextTime, duration] = (0, util_1.getTime)(Date.now(), this.options.flushInterval);
        const flushTick = (duration) => {
            if (this.closed)
                return;
            this.flushTicker = setTimeout(() => {
                this.flush(this.nextFlushTime).catch(util_1.noop);
                this.nextFlushTime += this.options.flushInterval;
                if (this.nextFlushTime < Date.now()) {
                    // 负载高的时候 nextFlushTime 可能比当前时间更早，纠正一下
                    const [nextTime, duration] = (0, util_1.getTime)(Date.now(), this.options.flushInterval);
                    this.nextFlushTime = nextTime;
                    flushTick(duration);
                }
                else {
                    flushTick(this.nextFlushTime - Date.now());
                }
                this.logger.info('Flush ticker is triggered, next flush time is', new Date(this.nextFlushTime).toLocaleTimeString());
            }, duration).unref();
        };
        this.nextFlushTime = nextTime;
        this.logger.info('Flush ticker has started, first flush time is', new Date(this.nextFlushTime).toLocaleTimeString());
        flushTick(duration);
    }
    /** flush metrics */
    async flush(sendTime = Date.now()) {
        let packedSize = 0;
        let packedCursorCount = 0;
        let packedMetricCount = 0;
        let totalCacheCursorCount = 0;
        let totalCacheMetricCount = 0;
        // pack all messages
        for (const [_, m] of this.metrics) {
            const summary = await this.flushMetric(m, sendTime);
            packedSize += summary.packedSize;
            packedCursorCount += summary.packedCursorCount;
            packedMetricCount += summary.packedMetricCount;
            totalCacheCursorCount += summary.cacheCursorCount;
            totalCacheMetricCount += summary.cacheMetricCount;
        }
        this.logger.info('Flush metrics finished, count: %s, size: %s, cost: %s, metric: %s(%s), cursor: %s(%s)', this.metrics.size, packedSize, Date.now() - sendTime, packedCursorCount, packedMetricCount, totalCacheCursorCount, totalCacheMetricCount);
        this.emit(consts_1.clientEvents.FLUSH);
        this.autoGC(totalCacheCursorCount, totalCacheMetricCount);
    }
    async flushMetric(m, sendTime = Date.now()) {
        return await m.matrix.flush(sendTime, async (packet) => {
            await this.send(packet).catch(util_1.noop);
        });
    }
    async send(packet) {
        if (!this.pool) {
            return;
        }
        for (let i = 0; i < this.requestTimes; i++) {
            let writer;
            try {
                writer = await this.pool.acquire();
                await writer.write(packet);
            }
            catch (e) {
                if (i >= this.requestTimes - 1) {
                    // retry failed
                    e.message = `Retry ${i} times but still failed: ${e.message}`;
                    this.logger.error(e);
                    throw e;
                }
                // retry after 1s
                await new Promise(resolve => setTimeout(resolve, 1000));
                continue;
            }
            finally {
                if (writer) {
                    // 这里必须 release ，不然 close 不掉
                    this.pool.release(writer);
                }
            }
            break;
        }
    }
    /** gc */
    gc(expiredDuration, ignoreMetrics) {
        const startTime = Date.now();
        const duration = expiredDuration !== null && expiredDuration !== void 0 ? expiredDuration : this.gcOptions.expireDuration;
        for (const [key, m] of this.metrics) {
            m.matrix.gc({
                ignoreMetrics,
                expiredDuration: duration,
                threshold: this.gcOptions.gcThreshold,
            });
            if (!m.keepAlive && !m.matrix.size) {
                // 非长缓存的 metric 在所有序列过期后也清理掉
                this.metrics.delete(key);
            }
        }
        this.logger.info('GC finished, cost: %sms', Date.now() - startTime);
        this.emit(consts_1.clientEvents.GC);
    }
    autoGC(totalCursor, totalMetric) {
        if (totalMetric > this.gcOptions.forceGCCacheMetricCount ||
            totalCursor > this.gcOptions.forceGCCacheCursorCount) {
            this.forceGCCount++;
            this.logger.info('Cache cursor/metric count is exceeded( cursor: %s / %s, metric: %s / %s ), trigger force gc( continuous gc count: %s )', totalCursor, this.gcOptions.forceGCCacheCursorCount, totalMetric, this.gcOptions.forceGCCacheMetricCount, this.forceGCCount);
            let ignoreCounter = true;
            // 连续强制 gc 次数超过一定次数打印一下缓存的状态，统计一下 counter 数量，判断是否把 counter 也 gc 掉
            if (this.forceGCCount > this.gcOptions.forceGCHealthCheckCount) {
                let counterMetricCount = 0;
                this.analyse((m, info) => {
                    var _a, _b;
                    this.logger.info('Metric %s cache info: %j', m.name, info);
                    counterMetricCount +=
                        (((_a = info.detail[metrics_codec_1.MetricType.Counter]) === null || _a === void 0 ? void 0 : _a.metricCount) || 0) +
                            (((_b = info.detail[metrics_codec_1.MetricType.DeltaCounterType]) === null || _b === void 0 ? void 0 : _b.metricCount) || 0);
                });
                // 如果单 counter 数量超了 forceGCCacheMetricCount，把 counter 也 gc 掉
                if (counterMetricCount > this.gcOptions.forceGCCacheMetricCount) {
                    this.logger.info('Cache counter metric count is exceed( %s / %s ), will gc counter', counterMetricCount, this.gcOptions.forceGCCacheMetricCount);
                    ignoreCounter = false;
                }
            }
            // 触发强制 gc，这里先绕过 counter，因为需要累加
            this.gc(0, ignoreCounter ? [metrics_codec_1.MetricType.Counter, metrics_codec_1.MetricType.DeltaCounterType] : undefined);
        }
        else {
            this.forceGCCount = 0;
        }
    }
    /** 获取当前缓存的信息 */
    analyse(cb) {
        const analyseInfo = {};
        for (const [_, m] of this.metrics) {
            const info = (analyseInfo[m.name] = analyseInfo[m.name] || {
                cursorCount: 0,
                metricCount: 0,
                activeMetricCount: 0,
                detail: {},
            });
            m.matrix.analyse(info);
            cb === null || cb === void 0 ? void 0 : cb(m, info);
        }
        return analyseInfo;
    }
    /**
     * 重置状态，主要是关闭连接，关闭计时器
     * 让 client 可以被回收，如果出现新的 metric，再恢复运作
     */
    async reset() {
        var _a;
        clearTimeout(this.flushTicker);
        clearInterval(this.gcTicker);
        this.metrics.clear();
        this.flushTicker = undefined;
        this.gcTicker = undefined;
        this.inited = false;
        await ((_a = this.pool) === null || _a === void 0 ? void 0 : _a.close());
    }
    async close() {
        await this.flush(this.nextFlushTime).catch(util_1.noop);
        await this.reset();
        this.emit(consts_1.clientEvents.CLOSE);
        this.removeAllListeners();
        this.closed = true;
    }
}
exports.Client = Client;
