// @byted-nodex/metrics/lib/consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.kTceSubClusterFile = exports.kTceClusterFile = exports.kOsPPEFile = exports.kMetricsEdgeConfFile = exports.defaultMeasureSuffix = exports.clientEvents = exports.defaultBuckets = exports.metricsTypesMethods = exports.candidateTimeIntervals = exports.defaultGCThreshold = exports.minGCInterval = exports.defaultGCInterval = exports.defaultExpireDuration = exports.defaultForceGCHealthCheckCount = exports.defaultForceGCCacheMetricCount = exports.defaultForceGCCursorCount = exports.defaultMaxCursorSize = exports.defaultShardingSize = exports.oneMessageSizeLimit = exports.defaultTimerFieldsSuffixes = exports.defaultDeltaSuffixes = exports.defaultHistoSuffixes = exports.bucketMaxCount = exports.DefaultClientSymbol = exports.maxASCIICode = exports.maxTagLen = void 0;
const metrics_codec_1 = require("@byted-nodex/metrics-codec");
exports.maxTagLen = 255;
exports.maxASCIICode = 127;
exports.DefaultClientSymbol = Symbol('MetricDefaultClient');
exports.bucketMaxCount = 100;
exports.defaultHistoSuffixes = ['sum', 'count'];
exports.defaultDeltaSuffixes = ['delta', 'rate', 'counter'];
exports.defaultTimerFieldsSuffixes = [
    'min',
    'max',
    'avg',
    'sum',
    'counter',
    'pct50',
    'pct90',
    'pct95',
    'pct99',
    'pct999',
];
/** message 最大体积 */
exports.oneMessageSizeLimit = (1 << 17) - 1;
exports.defaultShardingSize = 1024 * 1024;
/** 序列上限控制，超出后丢弃打点 */
exports.defaultMaxCursorSize = 500 * 10000;
/** 触发强制 gc 的阈值，超出后 flush 时强制 gc */
exports.defaultForceGCCursorCount = 20 * 10000;
exports.defaultForceGCCacheMetricCount = 100 * 10000;
exports.defaultForceGCHealthCheckCount = 3;
/** 序列过期时间为 24 小时 */
exports.defaultExpireDuration = 24 * 60 * 60 * 1000;
exports.defaultGCInterval = exports.defaultExpireDuration * 0.25;
exports.minGCInterval = 5 * 60 * 1000; // 最小 5min
exports.defaultGCThreshold = 0.3;
// 允许的精度列表
exports.candidateTimeIntervals = [1, 5, 15, 30, 60, 300, 600, 900].map(t => t * 1000);
exports.metricsTypesMethods = [
    [metrics_codec_1.MetricType.Store, 'emitStore'],
    [metrics_codec_1.MetricType.Counter, 'emitCounter'],
    [metrics_codec_1.MetricType.Timer, 'emitTimer'],
    [metrics_codec_1.MetricType.RateCounter, 'emitRateCounter'],
    [metrics_codec_1.MetricType.Meter, 'emitMeter'],
];
exports.defaultBuckets = [
    0.005, 0.01, 0.025, 0.05, 0.075, 0.1, 0.25, 0.5, 0.75, 1, 2.5, 5, 7.5, 10,
];
exports.clientEvents = {
    GC: 'gc',
    FLUSH: 'flush',
    CLOSE: 'close',
};
exports.defaultMeasureSuffix = {
    [metrics_codec_1.MetricType.RateCounter]: 'rate',
    [metrics_codec_1.MetricType.Timer]: 'timer',
    [metrics_codec_1.MetricType.Store]: 'store',
    [metrics_codec_1.MetricType.Meter]: 'meter',
    [metrics_codec_1.MetricType.Counter]: 'counter',
    [metrics_codec_1.MetricType.Histogram]: 'histogram',
    [metrics_codec_1.MetricType.DeltaCounterType]: 'delta_counter',
};
exports.kMetricsEdgeConfFile = '/opt/tmp/metric/edge.conf';
exports.kOsPPEFile = '/opt/tmp/byteos/ppe';
exports.kTceClusterFile = '/opt/tmp/tce/cluster';
exports.kTceSubClusterFile = '/opt/tmp/tce/subcluster';
