// @byted-nodex/metrics/lib/metric.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Metric = void 0;
const metrics_codec_1 = require("@byted-nodex/metrics-codec");
const collector_1 = require("./collector");
const fields_1 = require("./fields");
const util_1 = require("./util");
const consts_1 = require("./consts");
const noopEmitter = {
    emit() {
        // do nothing
    },
};
class Metric {
    constructor(options) {
        var _a;
        this.name = options.name;
        this.tenant = options.tenant;
        this.client = options.client;
        this.keepAlive = options.keepAlive !== false;
        this.cacheTagBytes = options.cacheTagBytes === true;
        this.forceMultiFields =
            typeof options.forceMultiFields === 'boolean'
                ? options.forceMultiFields
                : !!((_a = options.fields) === null || _a === void 0 ? void 0 : _a.length);
        // 指标的计算策略
        this.valuePolicy = {
            compatTimer: this.forceMultiFields === true
                ? false
                : typeof options.compatTimer === 'boolean'
                    ? options.compatTimer
                    : true,
            timerCacheCapacity: options.timerCacheCapacity || 2048,
            interval: options.interval || this.clientOps.flushInterval / 1000,
            reportLastValue: options.reportLastValue !== false,
            reportInitialValue: options.reportInitialValue !== false,
            reportUnchangeCounter: options.reportUnchangeCounter !== false,
        };
        const fields = [];
        // 强制多值，提供默认多值字段
        if (this.forceMultiFields) {
            fields.push((0, fields_1.newRateCounterField)(consts_1.defaultMeasureSuffix[metrics_codec_1.MetricType.RateCounter]), (0, fields_1.newCounterField)(consts_1.defaultMeasureSuffix[metrics_codec_1.MetricType.Counter]), (0, fields_1.newMeterField)(consts_1.defaultMeasureSuffix[metrics_codec_1.MetricType.Meter]), (0, fields_1.newStoreField)(consts_1.defaultMeasureSuffix[metrics_codec_1.MetricType.Store]), (0, fields_1.newTimerField)(consts_1.defaultMeasureSuffix[metrics_codec_1.MetricType.Timer]), (0, fields_1.newDeltaCounterField)(consts_1.defaultMeasureSuffix[metrics_codec_1.MetricType.DeltaCounterType]));
        }
        // 指定 buckets，开启 histogram 的多值
        if (options.buckets) {
            fields.push((0, fields_1.newHistogramField)(consts_1.defaultMeasureSuffix[metrics_codec_1.MetricType.Histogram], options.buckets));
        }
        /** 拓展 fields */
        this.fields = new fields_1.Fields(fields.concat(options.fields || []), this.forceMultiFields);
        const tagNames = options.tagNames;
        const globalTags = options.globalTags || {};
        const globalTagNames = (0, util_1.sortTagNames)(Object.keys(globalTags));
        /** 预先定义 tag 使用更高效的数据结构 */
        if (tagNames && tagNames.length) {
            this.sortedTags = (0, util_1.sortTagNames)(tagNames);
            this.sortedTags.forEach(name => {
                if (!(0, util_1.isValidString)(name)) {
                    throw new Error(`${name} is not a valid tag name`);
                }
            });
            // 提前排好序
            const allTagNames = (0, util_1.sortTagNames)((0, util_1.uniq)(this.sortedTags.concat(globalTagNames)));
            this.sortedUserTagPos = this.sortedTags.map(sn => allTagNames.findIndex(n => n === sn));
            this.sortedGlobalTagPos = globalTagNames.map(sn => allTagNames.findIndex(n => n === sn));
            this.allTagsLen = allTagNames.length;
            this.matrix = new collector_1.KdTreeCollector(this);
        }
        else {
            this.matrix = new collector_1.MapCollector(this);
        }
        this.globalTags = this.formatGlobalTags(globalTagNames, globalTags);
    }
    get clientOps() {
        return this.client.options;
    }
    get logger() {
        return this.client.logger;
    }
    formatGlobalTags(tagNames, globalTags) {
        return tagNames.map(tagName => {
            if (!(0, util_1.isValidString)(tagName)) {
                throw new Error(`${tagName} is not a valid tag name`);
            }
            return {
                key: tagName,
                value: (0, util_1.formatTagValue)(globalTags[tagName]),
            };
        });
    }
    emit(value, ...values) {
        return this.withTags().emit(value, ...values);
    }
    buildIndex(tags) {
        tags = tags || {};
        const tagNames = this.sortedTags || Object.keys(tags);
        const index = tagNames.map(name => (0, util_1.formatTagValue)(tags[name]));
        return {
            index,
            tagNames,
        };
    }
    withTags(tags) {
        const c = this.matrix.get(this.buildIndex(tags));
        if (!c)
            return noopEmitter;
        return c;
    }
    /** 将当前 metric 释放掉，后续新增的打点将不会再上报 */
    async close() {
        await this.client.releaseMetric(this);
    }
}
exports.Metric = Metric;
