// @universe-design/react/es/transfer/panel/Search.js
import _extends from "@babel/runtime/helpers/esm/extends";
import * as React from 'react';
import SearchOutlineOutlined from '@universe-design/icons-react/SearchOutlineOutlined';
import clsx from 'clsx';
import { useTransferContext } from '../hooks/TransferContext';
import { Input } from '../../input';
import { useLocale } from '../../config-provider';
import { sxSearch } from '../sx';
import { forwardHTMLRef } from '../../util/forwardRef';
export const TransferSearch = forwardHTMLRef((props, ref) => {
  const {
    className,
    onChange: onChangeProp,
    ...inputProps
  } = props;
  const {
    searchValue,
    setSearchValue,
    cssScope
  } = useTransferContext();
  const {
    global: {
      search
    }
  } = useLocale();
  const onChange = React.useCallback(e => {
    onChangeProp === null || onChangeProp === void 0 ? void 0 : onChangeProp(e);
    setSearchValue(e.target.value);
  }, [onChangeProp, setSearchValue]);
  return /*#__PURE__*/React.createElement(Input, _extends({
    className: clsx(className, cssScope('__search')),
    prefix: /*#__PURE__*/React.createElement(SearchOutlineOutlined, null),
    placeholder: search
  }, inputProps, {
    value: searchValue,
    onChange: onChange,
    sx: [sxSearch],
    ref: ref
  }));
});
