// @universe-design/react/es/transfer/panel/SourcePanel.js
import _extends from "@babel/runtime/helpers/esm/extends";
import stylex from '@universe-design/stylex';
import * as React from 'react';
import clsx from 'clsx';
import { useTransferContext } from '../hooks/TransferContext';
import { TransferSearch } from './Search';
import { Checkbox } from '../../checkbox';
import { TransferPanelHeader, PanelContainer } from './PanelContainer';
import { TransferSourceList } from './SourceList';
import { ListItemLike } from '../../shared/components/list';
import { useLocaleReceiver } from '../../config-provider';
import { forwardHTMLRef } from '../../util/forwardRef';
import { sxSelectAll } from '../sx';

/**
 * @public
 * `TransferSelectAll` 的类型。
 */

export const TransferSelectAll = forwardHTMLRef((props, ref) => {
  /** Context */
  const {
    cssScope,
    toggleSelectAll,
    isAllSelected,
    value,
    filteredOptions
  } = useTransferContext();

  /** Locale */
  const {
    selectAll
  } = useLocaleReceiver('transfer');
  if (filteredOptions.length === 0) {
    return null;
  }

  /** Props */
  const {
    label = (selected, total) => selectAll.f({
      selected: String(selected),
      total: String(total)
    }),
    className,
    ...htmlProps
  } = props;

  /** Checked status */
  let indeterminate = false;
  const checked = isAllSelected;
  if (!checked && value.length) {
    indeterminate = filteredOptions.some(op => value.includes(op.value));
  }
  return /*#__PURE__*/React.createElement("div", _extends({
    ref: ref,
    className: clsx(className, cssScope('__select-all'), stylex(sxSelectAll))
  }, htmlProps), /*#__PURE__*/React.createElement(ListItemLike, null, /*#__PURE__*/React.createElement(Checkbox, {
    indeterminate: indeterminate,
    checked: checked,
    onCheckChange: toggleSelectAll
  }, label(value.length, filteredOptions.length))));
});
export const TransferSourcePanelHeader = forwardHTMLRef((props, ref) => {
  /** Memo default children */
  const defaultChildren = React.useMemo(() => /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(TransferSearch, null), /*#__PURE__*/React.createElement(TransferSelectAll, null)), []);

  /** Props */
  const {
    children = defaultChildren,
    ...htmlProps
  } = props;
  return /*#__PURE__*/React.createElement(TransferPanelHeader, _extends({}, htmlProps, {
    ref: ref
  }), children);
});
export const TransferSourcePanel = forwardHTMLRef((props, ref) => {
  /** Memo default children */
  const defaultChildren = React.useMemo(() => /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(TransferSourcePanelHeader, null), /*#__PURE__*/React.createElement(TransferSourceList, null)), []);

  /** Props */
  const {
    children = defaultChildren,
    optionRenderer,
    emptyContent: noOptionContentProp,
    ...htmlProps
  } = props;

  /** Context */
  const {
    searchValue,
    cssScope
  } = useTransferContext();

  /** Locale */
  const {
    noSearchResult,
    noOptions
  } = useLocaleReceiver('transfer');
  const emptyContent = noOptionContentProp !== null && noOptionContentProp !== void 0 ? noOptionContentProp : searchValue ? noSearchResult : noOptions;
  return /*#__PURE__*/React.createElement(PanelContainer, _extends({}, htmlProps, {
    emptyContent: emptyContent,
    optionRenderer: optionRenderer,
    ref: ref,
    className: clsx(cssScope('__source-panel'), props.className)
  }), children);
});
