// @universe-design/react/es/transfer/panel/TargetPanel.js
import _extends from "@babel/runtime/helpers/esm/extends";
import stylex from '@universe-design/stylex';
import * as React from 'react';
import clsx from 'clsx';
import { Button } from '../../button';
import { useTransferContext } from '../hooks/TransferContext';
import { TransferPanelHeader, PanelContainer } from './PanelContainer';
import { TransferTargetList } from './TargetList';
import { ListItemLike } from '../../shared/components/list';
import { useLocaleReceiver } from '../../config-provider';
import { forwardHTMLRef } from '../../util/forwardRef';
import { sxClearButton, sxSelectionSummary } from '../sx';
export const TranferSelectionSummary = forwardHTMLRef((props, ref) => {
  /** Locale */
  const {
    selected,
    clear
  } = useLocaleReceiver('transfer');

  /** Props */
  const {
    showClearButton = true,
    selectionSummary = selectedCount => selected.f({
      selected: String(selectedCount)
    }),
    ...htmlProps
  } = props;

  /** Context */
  const {
    cssScope,
    value,
    clearAll
  } = useTransferContext();
  return /*#__PURE__*/React.createElement("div", _extends({
    className: clsx(stylex(sxSelectionSummary), cssScope('__selection-summary'))
  }, htmlProps, {
    ref: ref
  }), /*#__PURE__*/React.createElement(ListItemLike, null, selectionSummary(value.length)), showClearButton && value.length > 0 && /*#__PURE__*/React.createElement(Button, {
    type: "text",
    onClick: clearAll,
    color: "primary",
    className: stylex(sxClearButton)
  }, clear));
});
export const TransferTargetPanelHeader = forwardHTMLRef((props, ref) => {
  /** Memo default children */
  const defaultChildren = React.useMemo(() => /*#__PURE__*/React.createElement(TranferSelectionSummary, null), []);

  /** Props */
  const {
    children = defaultChildren,
    ...htmlProps
  } = props;
  return /*#__PURE__*/React.createElement(TransferPanelHeader, _extends({}, htmlProps, {
    ref: ref
  }), children);
});
export const TransferTargetPanel = forwardHTMLRef((props, ref) => {
  /** Memo default children */
  const defaultChildren = React.useMemo(() => /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(TransferTargetPanelHeader, null), /*#__PURE__*/React.createElement(TransferTargetList, null)), []);

  /** Props */
  const {
    children = defaultChildren,
    optionRenderer,
    emptyContent: noOptionContentProp,
    ...htmlProps
  } = props;

  /** Context */
  const {
    cssScope
  } = useTransferContext();

  /** Locale */
  const {
    noOptionsSelected
  } = useLocaleReceiver('transfer');
  return /*#__PURE__*/React.createElement(PanelContainer, _extends({}, htmlProps, {
    emptyContent: noOptionsSelected !== null && noOptionsSelected !== void 0 ? noOptionsSelected : noOptionContentProp,
    optionRenderer: optionRenderer,
    ref: ref,
    className: clsx(cssScope('__target-panel'), props.className)
  }), children);
});
