// @universe-design/react/es/transfer/panel/TransferTargetOption.js
import _extends from "@babel/runtime/helpers/esm/extends";
import CloseSmallOutlined from '@universe-design/icons-react/CloseSmallOutlined';
import stylex from '@universe-design/stylex';
import clsx from 'clsx';
import * as React from 'react';
import { Button } from '../../button';
import { ListItem, ListItemTitle } from '../../shared/components/list';
import { forwardHTMLRef } from '../../util/forwardRef';
import { useTransferContext } from '../hooks/TransferContext';
import { useTransferOptionContext } from '../hooks/TransferOptionContext';
import { sxRemoveButton, sxDisabledRemoveButton, sxOption } from '../sx';
import { optionLabel } from '../utils';
export const TransferTargetOptionRemoveButton = /*#__PURE__*/React.forwardRef((props, ref) => {
  const {
    option,
    toggle,
    disabled
  } = useTransferOptionContext();
  const {
    cssScope
  } = useTransferContext();
  const prefix = cssScope('__targetList');
  return /*#__PURE__*/React.createElement(Button, _extends({
    className: clsx(`${prefix}-remove`, stylex(stylex.compose(sxRemoveButton, disabled && sxDisabledRemoveButton))),
    type: "icon",
    size: "small",
    icon: /*#__PURE__*/React.createElement(CloseSmallOutlined, null),
    onClick: () => toggle(option.value),
    ref: ref
  }, props));
});
export const TransferTargetOptionLabel = forwardHTMLRef((props, ref) => {
  const {
    option
  } = useTransferOptionContext();
  return /*#__PURE__*/React.createElement(ListItemTitle, _extends({}, props, {
    ref: ref
  }), optionLabel(option));
});
export const TransferTargetOption = forwardHTMLRef((props, ref) => {
  var _props$sx;
  const defaultChildren = React.useMemo(() => /*#__PURE__*/React.createElement(TransferTargetOptionLabel, null), []);
  const defaultAfter = React.useMemo(() => /*#__PURE__*/React.createElement(TransferTargetOptionRemoveButton, null), []);
  const {
    children = defaultChildren,
    after = defaultAfter,
    ...restProps
  } = props;
  const {
    disabled
  } = useTransferOptionContext();
  return /*#__PURE__*/React.createElement(ListItem, _extends({
    disabled: disabled,
    after: after,
    ref: ref,
    showCheckmark: false
  }, restProps, {
    sx: [sxOption, ...((_props$sx = props.sx) !== null && _props$sx !== void 0 ? _props$sx : [])]
  }), children);
});
