// @universe-design/react/es/transfer/panel/TransferSourceOption.js
import _extends from "@babel/runtime/helpers/esm/extends";
import stylex from '@universe-design/stylex';
import clsx from 'clsx';
import * as React from 'react';
import { Checkbox } from '../../checkbox';
import { HighlightConsumer } from '../../shared/components/highlighter';
import { ListItem, ListItemTitle } from '../../shared/components/list';
import { TextOverflow } from '../../shared/components/text-overflow';
import { forwardHTMLRef } from '../../util/forwardRef';
import { composeEventHandlers } from '../../util/function';
import { useTransferContext } from '../hooks/TransferContext';
import { useTransferOptionContext } from '../hooks/TransferOptionContext';
import { sxTransferLabel, sxDisabledTransferCheckbox, sxOption, sxLabelHighlight } from '../sx';
import { optionLabel } from '../utils';
export const TransferSourceOptionCheckbox = /*#__PURE__*/React.forwardRef((props, ref) => {
  const {
    isChecked,
    disabled
  } = useTransferOptionContext();
  return /*#__PURE__*/React.createElement(Checkbox, _extends({}, props, {
    checked: isChecked,
    disabled: disabled,
    ref: ref
  }));
});
export const TransferSourceOptionLabel = forwardHTMLRef((props, ref) => {
  const {
    cssScope
  } = useTransferContext();
  const {
    option,
    disabled
  } = useTransferOptionContext();
  const markRenderer = React.useCallback(text => /*#__PURE__*/React.createElement("span", {
    className: clsx(!disabled && [cssScope('__label-highlight'), stylex(sxLabelHighlight)])
  }, text), [cssScope, disabled]);
  return /*#__PURE__*/React.createElement(ListItemTitle, _extends({
    textOverflow: false
  }, props, {
    ref: ref
  }), /*#__PURE__*/React.createElement(TextOverflow, {
    as: "div",
    className: clsx(stylex(sxTransferLabel), cssScope('__source-option-label'), disabled && [stylex(sxDisabledTransferCheckbox)]),
    lineClamp: 1
  }, /*#__PURE__*/React.createElement(HighlightConsumer, {
    markRenderer: markRenderer
  }, optionLabel(option))));
});
export const TransferSourceOption = forwardHTMLRef((props, ref) => {
  var _props$sx;
  const defaultChildren = React.useMemo(() => /*#__PURE__*/React.createElement(TransferSourceOptionLabel, null), []);
  const defaultBefore = React.useMemo(() => /*#__PURE__*/React.createElement(TransferSourceOptionCheckbox, null), []);
  const {
    children = defaultChildren,
    before = defaultBefore,
    ...restProps
  } = props;
  const {
    option,
    disabled,
    toggle
  } = useTransferOptionContext();
  composeEventHandlers();
  return /*#__PURE__*/React.createElement(ListItem, _extends({
    showCheckmark: false,
    disabled: disabled,
    before: before,
    ref: ref
  }, restProps, {
    sx: [sxOption, ...((_props$sx = props.sx) !== null && _props$sx !== void 0 ? _props$sx : [])],
    onClick: React.useMemo(() => composeEventHandlers(props.onClick, () => {
      if (disabled) {
        return;
      }
      toggle(option.value);
    }), [disabled, option.value, props.onClick, toggle])
  }), children);
});
