// @universe-design/react/src/rate/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import CBasic from '../demo/basic';
import CCharacterFunction from '../demo/character-function';
import CCharacter from '../demo/character';
import CClear from '../demo/clear';
import CHalf from '../demo/half';
import CNoDesc from '../demo/no-desc';
import CReadonly from '../demo/readonly';
import CText from '../demo/text';
if (!(CBasic as Story).storyName) {
  (CBasic as Story).storyName = '0. 基础用法';
  (CBasic as Story).parameters = {
    docs: {
      description: {
        story: '简单的用法。\n\n',
      },
    },
  };
}
if (!(CCharacterFunction as Story).storyName) {
  (CCharacterFunction as Story).storyName = '6. 自定义每个字符';
  (CCharacterFunction as Story).parameters = {
    docs: {
      description: {
        story: '通过`(RateProps) => ReactNode`的方式自定义每一个字符。\n\n',
      },
    },
  };
}
if (!(CCharacter as Story).storyName) {
  (CCharacter as Story).storyName = '5. 其他字符';
  (CCharacter as Story).parameters = {
    docs: {
      description: {
        story: '可以将星星替换为其他字符，比如字母、数字、字体图标甚至中文。\n\n',
      },
    },
  };
}
if (!(CClear as Story).storyName) {
  (CClear as Story).storyName = '4. 支持清除';
  (CClear as Story).parameters = {
    docs: {
      description: {
        story: '支持允许或者禁用清除。\n\n',
      },
    },
  };
}
if (!(CHalf as Story).storyName) {
  (CHalf as Story).storyName = '1. 半星';
  (CHalf as Story).parameters = {
    docs: {
      description: {
        story: '支持选中半星。\n\n',
      },
    },
  };
}
if (!(CNoDesc as Story).storyName) {
  (CNoDesc as Story).storyName = '2. 无评分描述';
  (CNoDesc as Story).parameters = {
    docs: {
      description: {
        story: '不显示描述信息。\n\n',
      },
    },
  };
}
if (!(CReadonly as Story).storyName) {
  (CReadonly as Story).storyName = '3. 只读';
  (CReadonly as Story).parameters = {
    docs: {
      description: {
        story: '只读，无法进行鼠标交互。\n\n',
      },
    },
  };
}
if (!(CText as Story).storyName) {
  (CText as Story).storyName = '2.1. 自定义文案';
  (CText as Story).parameters = {
    docs: {
      description: {
        story: '通过`descriptions`设置评分组件的默认描述文案，通过`tooltips`设置鼠标悬浮时的提示文案。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Rate/官网示例',
  tags: ['autodocs'],
};

export { CBasic, CCharacterFunction, CCharacter, CClear, CHalf, CNoDesc, CReadonly, CText };
