// @universe-design/react/lib/table/hooks/useSelection/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = useSelection;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _rcTable = require("@universe-design/rc-table");
var _difference = _interopRequireDefault(require("lodash/difference"));
var _without = _interopRequireDefault(require("lodash/without"));
var _reactUtilities = require("@universe-design/react-utilities");
var _SelectionTitle = _interopRequireDefault(require("./SelectionTitle"));
var _SelectionCell = _interopRequireDefault(require("./SelectionCell"));
var _constants = require("../../constants");
var _util = require("../../../util");
var _utils = require("../../utils");
var _tree = require("../../../tree");
function flattenData(data, childrenColumnName) {
  let list = [];
  if (!data) {
    return list;
  }
  data.forEach(record => {
    list.push(record);
    if (record && typeof record === 'object' && childrenColumnName in record) {
      const childrenData = record[childrenColumnName];
      list = [...list, ...flattenData(childrenData, childrenColumnName)];
    }
  });
  return list;
}
function useSelection(rowSelection, config) {
  const {
    selections = true,
    preserveSelectedRowKeys,
    selectedRowKeys,
    getCheckboxProps,
    onChange,
    onSelectAll,
    onSelectInvert,
    onSelect,
    columnWidth: selectionColWidth,
    type: selectionType = 'checkbox',
    renderCell,
    checkStrictly = true,
    hideSelectAll,
    fixed
  } = rowSelection || {};

  /**
   * 当没有设置 rowSelection 时，应该跳过 rowSelection 中耗时的操作
   */
  const shouldSkip = !rowSelection;
  const {
    prefix,
    tableLocale,
    data,
    pageData,
    getRecordByKey,
    getRowKey,
    getRowChildren,
    expandType,
    childrenColumnName
  } = config;
  const [mergedSelectedKeys, setMergedSelectedKeys] = (0, _reactUtilities.useControlledState)(selectedRowKeys || [], selectedRowKeys);
  const lastSelectedKey = _react.default.useRef(null);
  const recordsCacheRef = _react.default.useRef(new Map());
  const updateRecordsCache = (0, _react.useCallback)(keys => {
    if (preserveSelectedRowKeys) {
      const newCache = new Map();
      keys.forEach(key => {
        let record = getRecordByKey(key);
        if (!record && recordsCacheRef.current.has(key)) {
          record = recordsCacheRef.current.get(key);
        }
        newCache.set(key, record);
      });
      recordsCacheRef.current = newCache;
    } else {
      recordsCacheRef.current = new Map(keys.filter(getRecordByKey).map(k => [k, getRecordByKey(k)]));
    }
  }, [getRecordByKey, preserveSelectedRowKeys]);

  // update view when dataSource changed
  (0, _react.useEffect)(() => {
    if (shouldSkip) {
      return;
    }
    updateRecordsCache(mergedSelectedKeys);
    const newMergedSelectedKeys = [...recordsCacheRef.current.keys()];
    if ((0, _difference.default)(mergedSelectedKeys, newMergedSelectedKeys).length) {
      setMergedSelectedKeys(newMergedSelectedKeys);
    }
  }, [mergedSelectedKeys, setMergedSelectedKeys, updateRecordsCache, shouldSkip]);

  // Flatten Page Data
  const flattenedData = (0, _react.useMemo)(() => shouldSkip ? [] : flattenData(pageData, childrenColumnName), [pageData, childrenColumnName, shouldSkip]);

  // Checkbox Props Map
  const checkboxPropsMap = (0, _react.useMemo)(() => {
    const map = new Map();
    flattenedData.forEach(record => {
      const key = getRowKey(record);
      const checkboxProps = (getCheckboxProps ? getCheckboxProps(record) : null) || {};
      map.set(key, checkboxProps);
    });
    return map;
  }, [flattenedData, getRowKey, getCheckboxProps]);
  const {
    keyEntities
  } = (0, _react.useMemo)(() => shouldSkip || checkStrictly ? {
    keyEntities: null
  } : (0, _utils.convertDataToEntities)(data, getRowKey), [checkStrictly, data, getRowKey, shouldSkip]);
  // 不知道是干啥的。。。
  // const isCheckboxDisabled: GetCheckDisabled<RecordType> = useCallback(
  //   (r: RecordType) => Boolean(checkboxPropsMap.get(getRowKey(r))?.disabled),
  //   [checkboxPropsMap, getRowKey]
  // );

  const checkedCache = (0, _react.useMemo)(() => {
    if (checkStrictly || !keyEntities) {
      return new Map();
    }
    const cache = (0, _utils.checkedSet2Cache)(keyEntities, (0, _tree.array2Set)(mergedSelectedKeys));
    return cache;
  }, [checkStrictly, keyEntities, mergedSelectedKeys]);
  const [derivedSelectedKeys, derivedHalfSelectedKeys] = (0, _react.useMemo)(() => {
    if (checkStrictly || !keyEntities) {
      return [mergedSelectedKeys || [], []];
    }
    const keys = [];
    const halfKeys = [];
    for (const [key, value] of checkedCache.entries()) {
      if (value === _utils.CheckedType.Checked) {
        keys.push(key);
      }
      if (value === _utils.CheckedType.Indeterminate) {
        halfKeys.push(key);
      }
    }
    return [keys, halfKeys];
  }, [checkedCache, mergedSelectedKeys, checkStrictly, keyEntities]);
  const derivedSelectedKeySet = (0, _react.useMemo)(() => {
    const keys = selectionType === _constants.SELECTION_TYPE_RADIO ? derivedSelectedKeys.slice(0, 1) : derivedSelectedKeys;
    return new Set(keys);
  }, [derivedSelectedKeys, selectionType]);
  const derivedHalfSelectedKeySet = (0, _react.useMemo)(() => selectionType === _constants.SELECTION_TYPE_RADIO ? new Set() : new Set(derivedHalfSelectedKeys), [derivedHalfSelectedKeys, selectionType]);
  const setSelectedKeys = (0, _react.useCallback)(keys => {
    let availableKeys, records;
    updateRecordsCache(keys);
    if (preserveSelectedRowKeys) {
      // 用户设置了 preserveSelectedRowKeys 时，保留 key
      availableKeys = keys;
      records = keys.map(key => recordsCacheRef.current.get(key));
    } else {
      availableKeys = [];
      records = [];
      keys.forEach(key => {
        const record = getRecordByKey(key);
        if (record !== undefined) {
          availableKeys.push(key);
          records.push(record);
        }
      });
    }
    setMergedSelectedKeys(availableKeys, () => {
      (0, _util.isFunction)(onChange) && onChange(availableKeys, records);
    });
  }, [updateRecordsCache, preserveSelectedRowKeys, setMergedSelectedKeys, getRecordByKey, onChange]);

  // 用户手动勾选触发事件
  const triggerSelect = (0, _react.useCallback)((key, selected, keys, event) => {
    if (onSelect) {
      const rows = keys.map(k => getRecordByKey(k));
      onSelect(getRecordByKey(key), selected, rows, event);
    }
    setSelectedKeys(keys);
  }, [onSelect, getRecordByKey, setSelectedKeys]);
  const keySet = (0, _react.useMemo)(() => new Set(derivedSelectedKeySet), [derivedSelectedKeySet]);

  // 下拉菜单的配置项列表
  const mergedSelections = (0, _react.useMemo)(() => {
    if (!selections || hideSelectAll) {
      return null;
    }
    const selectionList = selections === true ? [_constants.SELECTION_ALL, _constants.SELECTION_INVERT] : selections;
    if (selectionList.length === 0) {
      return null;
    }
    return selectionList.map(selection => {
      if (selection === _constants.SELECTION_ALL) {
        return {
          key: 'all',
          text: tableLocale.selectionAll,
          onSelect() {
            setSelectedKeys(data.map(record => getRowKey(record)));
          }
        };
      }
      if (selection === _constants.SELECTION_INVERT) {
        return {
          key: 'invert',
          text: tableLocale.selectInvert,
          onSelect() {
            const invertSet = (_keySet, _pageData) => {
              const keys = [];
              const childrenKeys = [];
              _pageData.forEach(_data => {
                const key = getRowKey(_data);
                const children = getRowChildren === null || getRowChildren === void 0 ? void 0 : getRowChildren(_data);
                if (checkStrictly) {
                  !_keySet.has(key) && keys.push(key);
                  if (children) {
                    const [_keys] = invertSet(_keySet, children);
                    keys.push(..._keys);
                  }
                } else {
                  if (children) {
                    const [_keys, _childrenKeys] = invertSet(_keySet, children);
                    _childrenKeys.length === children.length && keys.push(key);
                    keys.push(..._keys);
                    childrenKeys.push(..._childrenKeys);
                  } else {
                    !_keySet.has(key) && keys.push(key);
                  }
                }
              });
              return [keys, childrenKeys];
            };
            const [keys] = invertSet(keySet, pageData);
            if (onSelectInvert) {
              onSelectInvert(keys);
            }
            /**
             * 反选本页，仅替换本页数据
             * flattenedData: 打平以后的本页数据
             * flattenedPageKeys: 打平以后的本页数据的所有 key
             * keySet: 更新前已被选择的所有 key
             * keys: 反选操作计算后本页应该被选择的 key
             * finalSelectedKeys：用更新前已被选择的所有 key 去处掉本页的key，再将本页更新后新选择的 key 加进去
             *
             * eg: 当前共两页数据：[[1, 2, 3], [4, 5, 6]]，已经选择[2, 3, 4]，当前在第二页，点击反选
             *  flattenedPageKeys: [4, 5, 6]
             *  keySet: [2, 3, 4]
             *  keys: [5, 6]
             *  finalSelectedKeys：[2, 3, 4] - [4, 5, 6] + [5, 6] = [2, 3, 5, 6]
             */
            const flattenedPageKeys = flattenedData.map(d => getRowKey(d));
            const finalSelectedKeys = [...(0, _without.default)([...keySet], ...flattenedPageKeys), ...keys];
            setSelectedKeys(finalSelectedKeys);
          }
        };
      }
      return selection;
    });
  }, [checkStrictly, selections, hideSelectAll, tableLocale.selectionAll, tableLocale.selectInvert, setSelectedKeys, data, getRowKey, getRowChildren, pageData, onSelectInvert, keySet, flattenedData]);
  const transformColumns = (0, _react.useCallback)(columns => {
    var _columns$;
    if (!rowSelection) {
      return columns;
    }
    const recordKeys = flattenedData.map(getRowKey).filter(key => !checkboxPropsMap.get(key).disabled);
    const checkedCurrentAll = recordKeys.every(key => keySet.has(key));
    const checkedCurrentSome = recordKeys.some(key => keySet.has(key));
    const allDisabled = flattenedData.every(record => {
      const key = getRowKey(record);
      const checkboxProps = checkboxPropsMap.get(key) || {};
      return checkboxProps.disabled;
    });
    let defaultTitle;
    if (selectionType === _constants.SELECTION_TYPE_CHECKBOX && !hideSelectAll) {
      defaultTitle = /*#__PURE__*/_react.default.createElement(_SelectionTitle.default, {
        prefixCls: prefix,
        checked: !allDisabled && Boolean(flattenedData.length) && checkedCurrentAll,
        disabled: flattenedData.length === 0 || allDisabled,
        indeterminate: !checkedCurrentAll && checkedCurrentSome,
        recordKeys: recordKeys,
        mergedSelections: mergedSelections,
        onSelectAllChange: ev => {
          const changeKeys = [];
          if (checkedCurrentAll) {
            recordKeys.forEach(key => {
              keySet.delete(key);
              changeKeys.push(key);
            });
          } else {
            recordKeys.forEach(key => {
              if (!keySet.has(key)) {
                keySet.add(key);
                changeKeys.push(key);
              }
            });
          }
          const keys = Array.from(keySet);
          setSelectedKeys(keys);
          if (onSelectAll) {
            const rows = keys.map(k => getRecordByKey(k));
            const changedRows = changeKeys.map(k => getRecordByKey(k));
            onSelectAll(ev.target.checked, rows, changedRows);
          }
        }
      });
    }
    const selectionColumn = {
      width: selectionColWidth,
      className: `${prefix}-selection-column`,
      title: rowSelection.columnTitle || defaultTitle,
      render: (_, record, index) => {
        var _checkboxProps$indete;
        const key = getRowKey(record);
        const checked = keySet.has(key);
        const checkboxProps = checkboxPropsMap.get(key);
        const indeterminate = derivedHalfSelectedKeySet.has(key);
        const mergedIndeterminate = expandType === 'nest' ? indeterminate : (_checkboxProps$indete = checkboxProps === null || checkboxProps === void 0 ? void 0 : checkboxProps.indeterminate) !== null && _checkboxProps$indete !== void 0 ? _checkboxProps$indete : indeterminate;
        let onCellSelect;
        if (selectionType === _constants.SELECTION_TYPE_RADIO) {
          onCellSelect = _ref => {
            let {
              nativeEvent
            } = _ref;
            if (!keySet.has(key)) {
              triggerSelect(key, true, [key], nativeEvent);
            }
          };
        } else {
          onCellSelect = e => {
            // 鼠标点击、键盘敲击 checkBox，收到 pointerEvent 事件
            // 是否会收到其他事件？
            const nativeEvent = e.nativeEvent;
            const shift = nativeEvent.shiftKey;
            if (checkStrictly) {
              if (shift) {
                // shift 连选
                const curSelectedKeys = (0, _utils.selectWithShift)(lastSelectedKey.current, key, recordKeys, keySet);
                triggerSelect(key, !checked, Array.from(curSelectedKeys), nativeEvent);
              } else {
                // 普通checkbox单选
                const curSelectedKeySet = new Set(keySet);
                if (checked) {
                  curSelectedKeySet.delete(key);
                } else {
                  curSelectedKeySet.add(key);
                }
                triggerSelect(key, !checked, Array.from(curSelectedKeySet), nativeEvent);
              }
            } else {
              // !checkStrictly
              if (!keyEntities) {
                return;
              }
              const entity = keyEntities[key];
              const checkedType = checkedCache.get(key);
              const checkedCacheClone = (0, _tree.cloneMap)(checkedCache);
              if (checkedType === _utils.CheckedType.Checked) {
                (0, _utils.unCheck)(entity, checkedCacheClone, checkStrictly !== null && checkStrictly !== void 0 ? checkStrictly : false, getRowKey);
              } else {
                (0, _utils.check)(entity, checkedCacheClone, checkStrictly !== null && checkStrictly !== void 0 ? checkStrictly : false, getRowKey);
              }
              const nextCheckedKeys = Object.entries((0, _utils.cache2CheckedSet)(keyEntities, checkedCacheClone)).filter(_ref2 => {
                let [, v] = _ref2;
                return v;
              }).map(_ref3 => {
                let [k] = _ref3;
                return k;
              });
              triggerSelect(key, !checked, nextCheckedKeys, nativeEvent);
            }
            lastSelectedKey.current = key;
          };
        }
        const cellNode = /*#__PURE__*/_react.default.createElement(_SelectionCell.default, (0, _extends2.default)({}, checkboxProps, {
          type: selectionType,
          checked: checked,
          indeterminate: mergedIndeterminate,
          onSelect: onCellSelect
        }));
        if (renderCell) {
          return renderCell(checked, record, index, cellNode);
        }
        return cellNode;
      },
      [_rcTable.INTERNAL_COL_DEFINE]: {
        className: `${prefix}-selection-col`
      }
    };
    return [{
      ...selectionColumn,
      fixed: fixed || ((_columns$ = columns[0]) === null || _columns$ === void 0 ? void 0 : _columns$.fixed)
    }, ...columns];
  }, [onSelectAll, rowSelection, flattenedData, getRowKey, selectionType, hideSelectAll, selectionColWidth, prefix, expandType, fixed, checkboxPropsMap, derivedHalfSelectedKeySet, renderCell, triggerSelect, checkStrictly, keyEntities, checkedCache, setSelectedKeys, getRecordByKey, mergedSelections, lastSelectedKey, keySet]);
  return [transformColumns, derivedSelectedKeySet];
}
