// @universe-design/react/lib/table/hooks/useSelection/SelectionTitle.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _dropdown = require("../../../dropdown");
var _expandButton = require("../../../shared/components/expand-button");
var _menu = require("../../../menu");
var _checkbox = require("../../../checkbox");
function SelectionTitle(props) {
  const {
    prefixCls,
    checked,
    disabled,
    indeterminate,
    getPopupContainer,
    recordKeys,
    onSelectAllChange,
    mergedSelections
  } = props;
  let customizeSelections;
  if (mergedSelections) {
    const menu = /*#__PURE__*/_react.default.createElement(_menu.Menu, {
      getPopupContainer: getPopupContainer
    }, mergedSelections.map((selection, index) => {
      const {
        key,
        text,
        onSelect
      } = selection;
      return /*#__PURE__*/_react.default.createElement(_menu.Menu.Item, {
        key: key || index,
        onClick: () => {
          if (onSelect) {
            onSelect(recordKeys);
          }
        }
      }, text);
    }));

    // eslint-disable-next-line react-hooks/rules-of-hooks
    const [visible, setVisible] = (0, _react.useState)(false);
    customizeSelections = /*#__PURE__*/_react.default.createElement(_dropdown.Dropdown, {
      visible: visible,
      onVisibleChange: e => setVisible(e),
      overlay: menu,
      getPopupContainer: getPopupContainer
    }, /*#__PURE__*/_react.default.createElement(_expandButton.ExpandButton, {
      isExpanded: visible,
      className: `${prefixCls}-selection-extra`,
      size: "xxs",
      iconType: "DownBoldOutlined"
    }));
  }
  return /*#__PURE__*/_react.default.createElement("div", {
    className: `${prefixCls}-selection`
  }, /*#__PURE__*/_react.default.createElement(_checkbox.Checkbox, {
    checked: checked,
    indeterminate: indeterminate,
    onChange: onSelectAllChange,
    disabled: disabled
  }), customizeSelections);
}
var _default = exports.default = SelectionTitle;
