// @universe-design/react/src/space/index.tsx
import clsx from 'clsx';
import type { HTMLAttributes, Ref } from 'react';
import React, { forwardRef } from 'react';
import { useCssScope } from '../config-provider/hooks';
import type { DefaultSizeType } from '../util';
import { toFlatArray } from '../util';

import './style';

/**
 * 间距尺寸
 * @public
 */
export type SpaceSize = number | DefaultSizeType;

const spaceSize = {
  small: 8,
  medium: 16,
  large: 24,
};

function getNumberSize(size: SpaceSize) {
  return typeof size === 'string' ? spaceSize[size] : size || 0;
}
/**
 * @public
 */
export interface SpaceProps extends Omit<HTMLAttributes<HTMLDivElement>, 'className' | 'style'> {
  /**
   * `className` 属性透传
   */
  className?: string;
  /**
   * `style` 属性透传
   */
  style?: React.CSSProperties;
  /**
   * 间距大小
   * @defaultValue `"small"`
   */
  size?: SpaceSize | [SpaceSize, SpaceSize];
  /**
   * 对齐方式
   */
  align?: 'start' | 'end' | 'center' | 'baseline' | 'stretch';
  /**
   * 间距方向
   * @defaultValue `"horizontal"`
   */
  direction?: 'horizontal' | 'vertical';
  /**
   * 设置拆分
   */
  split?: React.ReactNode;
  /**
   * 是否自动换行，仅在 `horizontal` 时有效
   * @defaultValue `false`
   */
  wrap?: boolean;
  /**
   * 是否保留空项目
   * @defaultValue `false`
   */
  keepEmpty?: boolean;
  /**
   * 是否为 `inline` 元素
   * @defaultValue `true`
   */
  inline?: boolean;
  /**
   * @internal
   * flex item 的 min-width 默认为 auto，这会使块元素占用比预期更多的空间，甚至在小屏幕上将其容器拉伸到屏幕边缘之外。这个属性可以控制子元素是否应该超出 flex 容器
   * @defaultValue false
   */
  shouldItemConfinedToContainer?: boolean;
  /**
   * 我们默认使用数组 index 作为 item key，但是在某些动态 items 场景下会出现节点复用导致的问题，该属性可以用户自定义 item key。
   */
  itemKey?: (index: number) => React.Key;
  /**
   * @internal
   * 是否在子元素外包裹一个 display-inside: flow-root 元素，让子元素处于 BFC 中。当为 false 时，`shouldItemConfinedToContainer` 会无效。
   * @defaultValue true
   */
  shouldPlaceItemInFlowRoot?: boolean;
}

export const Space = forwardRef(function Space(
  {
    children,
    align,
    size = 'small',
    direction = 'horizontal',
    className,
    wrap = false,
    split,
    style,
    keepEmpty = false,
    inline = true,
    shouldItemConfinedToContainer = false,
    itemKey,
    shouldPlaceItemInFlowRoot = true,
    ...restProps
  }: SpaceProps,
  ref: Ref<HTMLDivElement>,
) {
  const cssScope = useCssScope('space');
  const itemCssScope = useCssScope('space__item');
  const childNodes = toFlatArray(children, { keepEmpty });
  const [horizontalSize, verticalSize] = React.useMemo(
    () => (Array.isArray(size) ? size : ([size, size] as [SpaceSize, SpaceSize])).map(getNumberSize),
    [size],
  );
  const latestIndex = childNodes.length - 1;

  if (childNodes.length === 0) {
    return null;
  }

  const mergedAlign = (() => {
    if (align) {
      return align;
    }
    if (direction === 'horizontal') {
      return 'center';
    }
    if (!shouldPlaceItemInFlowRoot) {
      return 'start';
    }
  })();

  return (
    <div
      {...restProps}
      ref={ref}
      className={clsx(
        cssScope(),
        cssScope(`-${direction}`),
        inline && cssScope('-inline'),
        mergedAlign && cssScope(`-align-${mergedAlign}`),
        className,
      )}
      style={{
        ...(wrap && { flexWrap: 'wrap' }),
        columnGap: horizontalSize / (split ? 2 : 1),
        rowGap: verticalSize,
        ...style,
      }}
    >
      {childNodes.map((node, index) => (
        <React.Fragment key={itemKey ? itemKey(index) : index}>
          {shouldPlaceItemInFlowRoot ? (
            <div className={clsx(itemCssScope(), shouldItemConfinedToContainer && itemCssScope('-confined'))}>
              {node}
            </div>
          ) : (
            node
          )}
          {index < latestIndex && split && <span className={itemCssScope('__split')}>{split}</span>}
        </React.Fragment>
      ))}
    </div>
  );
});
