// @universe-design/styles/ts/universe-design/index.dark.atom.ts
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
export const udtoken_AI_cot_shinytext_mask = 'rgba(0, 0, 0, 0.5)';
export const udtoken_block_view_inline_bg_nopermission = 'rgba(235, 235, 235, 0.1)';
export const udtoken_block_view_inline_bg_other = 'rgba(38, 85, 182, 0.35)';
export const udtoken_block_view_inline_bg_self = '#75a4ff';
export const udtoken_btn_ghost_bg = 'rgba(10, 10, 10, 0)';
export const udtoken_btn_pri_text_disabled = 'rgba(235, 235, 235, 0.3)';
export const udtoken_btn_se_bg_danger_focus = 'rgba(240, 91, 86, 0.2)';
export const udtoken_btn_se_bg_danger_hover = 'rgba(240, 91, 86, 0.2)';
export const udtoken_btn_se_bg_danger_pressed = 'rgba(240, 91, 86, 0.3)';
export const udtoken_btn_se_bg_neutral_focus = 'rgba(235, 235, 235, 0.1)';
export const udtoken_btn_se_bg_neutral_hover = 'rgba(235, 235, 235, 0.1)';
export const udtoken_btn_se_bg_neutral_pressed = 'rgba(235, 235, 235, 0.2)';
export const udtoken_btn_se_bg_pri_focus = 'rgba(76, 136, 255, 0.2)';
export const udtoken_btn_se_bg_pri_hover = 'rgba(76, 136, 255, 0.2)';
export const udtoken_btn_se_bg_pri_pressed = 'rgba(76, 136, 255, 0.3)';
export const udtoken_btn_se_bg_success_focus = 'rgba(65, 158, 52, 0.2)';
export const udtoken_btn_se_bg_success_hover = 'rgba(65, 158, 52, 0.2)';
export const udtoken_btn_se_bg_success_pressed = 'rgba(65, 158, 52, 0.3)';
export const udtoken_btn_selected_bg_danger_hover = 'rgba(240, 91, 86, 0.3)';
export const udtoken_btn_selected_bg_danger_normal = 'rgba(240, 91, 86, 0.15)';
export const udtoken_btn_selected_bg_danger_press = 'rgba(240, 91, 86, 0.3)';
export const udtoken_btn_selected_bg_neutral_hover = 'rgba(235, 235, 235, 0.12)';
export const udtoken_btn_selected_bg_neutral_normal = 'rgba(235, 235, 235, 0.08)';
export const udtoken_btn_selected_bg_neutral_press = 'rgba(235, 235, 235, 0.16)';
export const udtoken_btn_selected_bg_success_hover = 'rgba(65, 158, 52, 0.2)';
export const udtoken_btn_selected_bg_success_normal = 'rgba(65, 158, 52, 0.15)';
export const udtoken_btn_selected_bg_success_press = 'rgba(65, 158, 52, 0.3)';
export const udtoken_btn_selected_border_neutral = 'rgba(255, 255, 255, 0.5)';
export const udtoken_btn_selected_text_danger = '#f05b56';
export const udtoken_btn_selected_text_success = '#419e34';
export const udtoken_btn_text_bg_danger_focus = 'rgba(240, 91, 86, 0.2)';
export const udtoken_btn_text_bg_danger_hover = 'rgba(240, 91, 86, 0.2)';
export const udtoken_btn_text_bg_danger_pressed = 'rgba(240, 91, 86, 0.3)';
export const udtoken_btn_text_bg_neutral_focus = 'rgba(235, 235, 235, 0.1)';
export const udtoken_btn_text_bg_neutral_hover = 'rgba(235, 235, 235, 0.1)';
export const udtoken_btn_text_bg_neutral_pressed = 'rgba(235, 235, 235, 0.2)';
export const udtoken_btn_text_bg_pri_focus = 'rgba(76, 136, 255, 0.2)';
export const udtoken_btn_text_bg_pri_hover = 'rgba(76, 136, 255, 0.2)';
export const udtoken_btn_text_bg_pri_pressed = 'rgba(76, 136, 255, 0.3)';
export const udtoken_chatsender_stopgeneration_bg1_hover = '#505050';
export const udtoken_chatsender_stopgeneration_bg1_normal = '#424242';
export const udtoken_chatsender_stopgeneration_bg1_pressed = '#454545';
export const udtoken_chatsender_stopgeneration_bg2_normal = '#f5f5f5';
export const udtoken_colorpicker_blue = '#4c88ff';
export const udtoken_colorpicker_carmine = '#c24a8e';
export const udtoken_colorpicker_green = '#51ba43';
export const udtoken_colorpicker_indigo = '#5e68e8';
export const udtoken_colorpicker_lime = '#a6c313';
export const udtoken_colorpicker_neutral = '#757575';
export const udtoken_colorpicker_orange = '#f89e44';
export const udtoken_colorpicker_purple = '#8c55ec';
export const udtoken_colorpicker_red = '#f05b56';
export const udtoken_colorpicker_sunflower = '#f7e663';
export const udtoken_colorpicker_turquoise = '#17cfb5';
export const udtoken_colorpicker_violet = '#d661d6';
export const udtoken_colorpicker_wathet = '#51c3f0';
export const udtoken_colorpicker_yellow = '#fbcb46';
export const udtoken_component_bg_bottom_mask = 'linear-gradient(to bottom, rgba(26, 26, 26, 0) 0%, #1a1a1a 100%)';
export const udtoken_component_bg_left_mask = 'linear-gradient(to left, rgba(26, 26, 26, 0) 0%, #1a1a1a 100%)';
export const udtoken_component_bg_right_mask = 'linear-gradient(to right, rgba(26, 26, 26, 0) 0%, #1a1a1a 100%)';
export const udtoken_component_bg_top_mask = 'linear-gradient(to top, rgba(26, 26, 26, 0) 0%, #1a1a1a 100%)';
export const udtoken_component_outlined_bg = 'rgba(10, 10, 10, 0)';
export const udtoken_component_text_disabled_loading = '#757575';
export const udtoken_dropdown_button_divider = '#505050';
export const udtoken_imageview_bg_mask = 'rgba(0, 0, 0, 0.8)';
export const udtoken_imageview_icondisabled = 'rgba(255, 255, 255, 0.6)';
export const udtoken_imageview_toolbox_bg = 'rgba(41, 41, 41, 0.8)';
export const udtoken_imageview_tooltips_bg = '#505050';
export const udtoken_input_bg_disabled = 'rgba(235, 235, 235, 0.12)';
export const udtoken_message_card_bg_blue = '#173166';
export const udtoken_message_card_bg_carmine = '#591c3f';
export const udtoken_message_card_bg_green = '#173b12';
export const udtoken_message_card_bg_indigo = '#2a2d69';
export const udtoken_message_card_bg_light_neutral = '#373737';
export const udtoken_message_card_bg_lime = '#303804';
export const udtoken_message_card_bg_mask_general = 'rgba(10, 10, 10, 0.3)';
export const udtoken_message_card_bg_mask_special = 'rgba(10, 10, 10, 0.3)';
export const udtoken_message_card_bg_neutral = '#434343';
export const udtoken_message_card_bg_orange = '#4a2b10';
export const udtoken_message_card_bg_purple = '#3f2073';
export const udtoken_message_card_bg_red = '#591f1d';
export const udtoken_message_card_bg_turquoise = '#173b36';
export const udtoken_message_card_bg_violet = '#541854';
export const udtoken_message_card_bg_wathet = '#103647';
export const udtoken_message_card_bg_yellow = '#473409';
export const udtoken_message_card_text_blue = '#8fb4ff';
export const udtoken_message_card_text_carmine = '#fc94cf';
export const udtoken_message_card_text_green = '#69cc5c';
export const udtoken_message_card_text_indigo = '#aaaff8';
export const udtoken_message_card_text_lime = '#a6c313';
export const udtoken_message_card_text_neutral = '#cfcfcf';
export const udtoken_message_card_text_orange = '#f89e44';
export const udtoken_message_card_text_purple = '#c5a3ff';
export const udtoken_message_card_text_red = '#f89e9b';
export const udtoken_message_card_text_turquoise = '#17cfb5';
export const udtoken_message_card_text_violet = '#e99be9';
export const udtoken_message_card_text_wathet = '#51c3f0';
export const udtoken_message_card_text_yellow = '#fcd456';
export const udtoken_navigation_bar_resize_handle = '#4599ff';
export const udtoken_navigation_bar_tab_bg_black = '#0d0e12';
export const udtoken_navigation_bar_tab_bg_blue = '#141c2a';
export const udtoken_navigation_bar_tab_bg_gray = '#242424';
export const udtoken_progress_bg = 'rgba(235, 235, 235, 0.2)';
export const udtoken_quote_bar_bg = '#5f5f5f';
export const udtoken_reaction_bg_grey = '#101010';
export const udtoken_reaction_bg_grey_float = '#1f1f1f';
export const udtoken_side_navigation_fill_focus = 'rgba(235, 235, 235, 0.12)';
export const udtoken_side_navigation_fill_hover = 'rgba(235, 235, 235, 0.05)';
export const udtoken_side_navigation_fill_pressed = 'rgba(235, 235, 235, 0.12)';
export const udtoken_side_navigation_fill_selected = 'rgba(235, 235, 235, 0.1)';
export const udtoken_skeleton_bg = 'rgba(240, 240, 240, 0.08)';
export const udtoken_skeleton_fg = 'rgba(240, 240, 240, 0.05)';
export const udtoken_sliding_block_bg_disabled_loading = '#757575';
export const udtoken_steps_bg_hover = 'rgba(76, 136, 255, 0.15)';
export const udtoken_switch_handle_disabled = 'rgba(255, 255, 255, 0.3)';
export const udtoken_tab_pri_bg = '#194294';
export const udtoken_tab_pri_text = '#4c88ff';
export const udtoken_tab_se_bg_unselected = 'rgba(235, 235, 235, 0.2)';
export const udtoken_table_bg_expand = '#0a0a0a';
export const udtoken_table_bg_grey = '#292929';
export const udtoken_table_bg_head = '#373737';
export const udtoken_table_bg_hover = '#434343';
export const udtoken_table_bg_pressed = '#5f5f5f';
export const udtoken_table_bg_selected = '#173166';
export const udtoken_tabs_line_border_card = '#434343';
export const udtoken_tabs_view_bg_hover = 'rgba(235, 235, 235, 0.05)';
export const udtoken_tag_AI_trial_bg_normal =
  'radial-gradient(100.62% 141.42% at 100% 0%, rgba(124, 74, 212, 0.18) 0%, rgba(123, 131, 247, 0.18) 79.5%)';
export const udtoken_tag_AI_trial_bg_text_normal =
  'radial-gradient(58.07% 105.94% at 95% 28.12%, #a575fa 0%, #9499f7 100%)';
export const udtoken_tag_bg_blue = 'rgba(76, 136, 255, 0.2)';
export const udtoken_tag_bg_blue_default = 'rgba(117, 164, 255, 0.12)';
export const udtoken_tag_bg_blue_hover = 'rgba(117, 164, 255, 0.3)';
export const udtoken_tag_bg_blue_pressed = 'rgba(117, 164, 255, 0.2)';
export const udtoken_tag_bg_blue_solid = '#173166';
export const udtoken_tag_bg_carmine = 'rgba(219, 94, 164, 0.2)';
export const udtoken_tag_bg_carmine_hover = 'rgba(219, 94, 164, 0.3)';
export const udtoken_tag_bg_carmine_solid = '#591c3f';
export const udtoken_tag_bg_green = 'rgba(65, 158, 52, 0.2)';
export const udtoken_tag_bg_green_hover = 'rgba(65, 158, 52, 0.3)';
export const udtoken_tag_bg_green_solid = '#173b12';
export const udtoken_tag_bg_indigo = 'rgba(123, 131, 247, 0.2)';
export const udtoken_tag_bg_indigo_hover = 'rgba(123, 131, 247, 0.3)';
export const udtoken_tag_bg_indigo_solid = '#2a2d69';
export const udtoken_tag_bg_lime = 'rgba(130, 153, 10, 0.2)';
export const udtoken_tag_bg_lime_hover = 'rgba(130, 153, 10, 0.3)';
export const udtoken_tag_bg_lime_solid = '#303804';
export const udtoken_tag_bg_orange = 'rgba(219, 112, 24, 0.2)';
export const udtoken_tag_bg_orange_hover = 'rgba(219, 112, 24, 0.3)';
export const udtoken_tag_bg_orange_pressed = 'rgba(219, 112, 24, 0.35)';
export const udtoken_tag_bg_orange_solid = '#4a2b10';
export const udtoken_tag_bg_purple = 'rgba(165, 117, 250, 0.2)';
export const udtoken_tag_bg_purple_hover = 'rgba(165, 117, 250, 0.3)';
export const udtoken_tag_bg_purple_solid = '#3f2073';
export const udtoken_tag_bg_red = 'rgba(209, 70, 66, 0.2)';
export const udtoken_tag_bg_red_hover = 'rgba(209, 70, 66, 0.3)';
export const udtoken_tag_bg_red_solid = '#591f1d';
export const udtoken_tag_bg_turquoise = 'rgba(31, 161, 143, 0.2)';
export const udtoken_tag_bg_turquoise_hover = 'rgba(31, 161, 143, 0.3)';
export const udtoken_tag_bg_turquoise_solid = '#173b36';
export const udtoken_tag_bg_violet = 'rgba(214, 97, 214, 0.2)';
export const udtoken_tag_bg_violet_hover = 'rgba(214, 97, 214, 0.3)';
export const udtoken_tag_bg_violet_solid = '#541854';
export const udtoken_tag_bg_wathet = 'rgba(16, 153, 204, 0.2)';
export const udtoken_tag_bg_wathet_hover = 'rgba(16, 153, 204, 0.3)';
export const udtoken_tag_bg_wathet_solid = '#103647';
export const udtoken_tag_bg_yellow = 'rgba(212, 155, 11, 0.15)';
export const udtoken_tag_bg_yellow_hover = 'rgba(240, 182, 34, 0.3)';
export const udtoken_tag_bg_yellow_solid = '#473409';
export const udtoken_tag_close_blue_disabled = 'rgba(117, 164, 255, 0.8)';
export const udtoken_tag_close_blue_hover = '#75a4ff';
export const udtoken_tag_close_blue_normal = 'rgba(117, 164, 255, 0.6)';
export const udtoken_tag_close_blue_pressed = 'rgba(117, 164, 255, 0.8)';
export const udtoken_tag_close_orange_disabled = 'rgba(219, 112, 24, 0.3)';
export const udtoken_tag_close_orange_hover = '#db7018';
export const udtoken_tag_close_orange_normal = 'rgba(219, 112, 24, 0.6)';
export const udtoken_tag_close_orange_pressed = 'rgba(219, 112, 24, 0.8)';
export const udtoken_tag_neutral_bg_inverse = '#5f5f5f';
export const udtoken_tag_neutral_bg_inverse_opacity = 'rgba(235, 235, 235, 0.4)';
export const udtoken_tag_neutral_bg_normal = 'rgba(166, 166, 166, 0.2)';
export const udtoken_tag_neutral_bg_normal_hover = 'rgba(235, 235, 235, 0.3)';
export const udtoken_tag_neutral_bg_normal_pressed = 'rgba(235, 235, 235, 0.4)';
export const udtoken_tag_neutral_bg_solid = '#373737';
export const udtoken_tag_neutral_close_disabled = 'rgba(235, 235, 235, 0.3)';
export const udtoken_tag_neutral_close_hover = '#ebebeb';
export const udtoken_tag_neutral_close_normal = 'rgba(235, 235, 235, 0.6)';
export const udtoken_tag_neutral_close_pressed = 'rgba(235, 235, 235, 0.8)';
export const udtoken_tag_neutral_text_inverse = '#ffffff';
export const udtoken_tag_neutral_text_inverse_opacity = '#ffffff';
export const udtoken_tag_neutral_text_normal = '#a6a6a6';
export const udtoken_tag_neutral_text_solid = '#a6a6a6';
export const udtoken_tag_state_blue_light = '#4c88ff';
export const udtoken_tag_state_green_light = '#419e34';
export const udtoken_tag_state_grey_light = '#757575';
export const udtoken_tag_state_orange_light = '#db7018';
export const udtoken_tag_state_purple_light = '#a575fa';
export const udtoken_tag_state_red_light = '#f05b56';
export const udtoken_tag_text_blue = '#8fb4ff';
export const udtoken_tag_text_carmine = '#fc94cf';
export const udtoken_tag_text_green = '#69cc5c';
export const udtoken_tag_text_indigo = '#aaaff8';
export const udtoken_tag_text_lime = '#a6c313';
export const udtoken_tag_text_orange = '#f89e44';
export const udtoken_tag_text_purple = '#c5a3ff';
export const udtoken_tag_text_red = '#f89e9b';
export const udtoken_tag_text_s_blue = '#8fb4ff';
export const udtoken_tag_text_s_carmine = '#fc94cf';
export const udtoken_tag_text_s_green = '#69cc5c';
export const udtoken_tag_text_s_indigo = '#aaaff8';
export const udtoken_tag_text_s_lime = '#a6c313';
export const udtoken_tag_text_s_orange = '#f89e44';
export const udtoken_tag_text_s_purple = '#c5a3ff';
export const udtoken_tag_text_s_red = '#f89e9b';
export const udtoken_tag_text_s_turquoise = '#17cfb5';
export const udtoken_tag_text_s_violet = '#e99be9';
export const udtoken_tag_text_s_wathet = '#51c3f0';
export const udtoken_tag_text_s_yellow = '#fcd456';
export const udtoken_tag_text_turquoise = '#17cfb5';
export const udtoken_tag_text_violet = '#e99be9';
export const udtoken_tag_text_wathet = '#51c3f0';
export const udtoken_tag_text_yellow = '#fcd456';
export const udtoken_upload_bg_error = 'rgba(235, 235, 235, 0.12)';
export const udtoken_upload_mask_img = 'rgba(0, 0, 0, 0.7)';
