// @byted-service/bytedtrace/lib/span/consumerSpanInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConsumerSpanInfo = void 0;
const consts_1 = require("../consts");
const extInfo_1 = require("./extInfo");
class ConsumerSpanInfo extends extInfo_1.ExtInfo {
    constructor() {
        super();
        this.consumerGroup = '-';
        this.producerService = '-';
        this.producerCluster = '-';
        this.producerDc = '-';
        this.mqPartition = '-';
    }
    appendExtraSpanTags(tags) {
        tags.push([consts_1.CONSUMER_GROUP_KEY, this.consumerGroup], [consts_1.PRODUCER_SERVICE_KEY, this.producerService], [consts_1.PRODUCER_ADDR_KEY, this.producerAddr], [consts_1.PRODUCER_DC_KEY, this.producerDc], [consts_1.MQ_PARTITION_KEY, this.mqPartition], [consts_1.PRODUCER_CLUSTER_KEY, this.producerCluster]);
        if (this.produceTs) {
            tags.push([consts_1.PRODUCE_TS_KEY, this.produceTs * 1000]);
        }
        if (this.mqLag) {
            tags.push([consts_1.MQ_LAG_KEY, this.mqLag * 1000]);
        }
        return tags;
    }
}
exports.ConsumerSpanInfo = ConsumerSpanInfo;
