// @byted-service/bytedtrace/lib/span/spanContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpanContext = void 0;
const tslib_1 = require("tslib");
const gConsts = tslib_1.__importStar(require("../consts"));
const gUtils = tslib_1.__importStar(require("../utils"));
const tracer_1 = require("../tracer");
class SpanContext {
    constructor() {
        /**
         * SpanID represents span ID that must be unique within its trace,
         * but does not have to be globally unique.
         * 十进制数字字符串
         */
        this.spanID = '0';
        /**
         * ParentID refers to the ID of the parent span.
         * Should be 0 if the current span is a root span.
         * 十进制数字字符串
         */
        this.parentID = '0';
        /**
         * Flags is a bitmap containing such bits as 'sampled' and 'isDebug'.
         */
        this.flags = 0;
        /**
         * For example: dyeing
         */
        this.debugInfo = '';
        this.traceID = new tracer_1.TraceId(0, 0);
    }
    /**
     * 如果 carrier 中有 x-tt-trace 那么调用 bytedExtract 得到的 SpanContext 调用此方法返回 true 否则返回 false。
     * @returns
     */
    isDebug() {
        return this.debugInfo !== '';
    }
    /**
     * 如果是 NoopSpanContext 的实例，调用此方法返回 true，否则返回 false
     * @returns
     */
    noop() {
        return false;
    }
    toStringJaeger() {
        // 注入到 carrier 必须是16进制
        const spanID = gUtils.intStrTo16HexStr(this.spanID);
        const parentID = gUtils.intStrTo16HexStr(this.parentID);
        const high = gUtils.intStrTo16HexStr(this.traceID.high);
        const low = gUtils.intStrTo16HexStr(this.traceID.low);
        if (this.traceID.high === '0' || !this.traceID.high) {
            return `${low}:${spanID}:${parentID}:${this.flags}`;
        }
        const traceID = high + gUtils.fillZeroOnHigh(16, low);
        return `${traceID}:${spanID}:${parentID}:${this.flags}`;
    }
    toStringW3C() {
        const version = gUtils.fillZeroOnHigh(2, gConsts.GLOBAL_TRACE_VERSION);
        const high = gUtils.intStrTo16HexStr(this.traceID.high);
        const low = gUtils.intStrTo16HexStr(this.traceID.low);
        const traceID = gUtils.fillZeroOnHigh(16, high) + gUtils.fillZeroOnHigh(16, low);
        const spanId = gUtils.fillZeroOnHigh(16, gUtils.intStrTo16HexStr(this.spanID));
        const flags = gUtils.fillZeroOnHigh(2, this.flags);
        return `${version}-${traceID}-${spanId}-${flags}`;
    }
    /**
     * 获取 baggage 中键值为 key 的 value 值，由 Span.setBaggageItem 设置，但是 SpanContext 中的 baggage 是从 Span 中拷贝而来。
     * @param key
     * @returns
     */
    getBaggageItem(key) {
        if (typeof key !== 'string' || key === '') {
            return '';
        }
        return this.baggage.get(key) || '';
    }
    /**
     * 判断是否是有效的 SpanContext，返回值为：this.debugInfo !== undefined || !this.noop() && this.traceID && this.traceID.isValid();
     * @returns
     */
    isValid() {
        return (this.debugInfo !== undefined ||
            (!this.noop() && this.traceID && this.traceID.isValid()));
    }
}
exports.SpanContext = SpanContext;
