// @byted-service/bytedtrace/lib/tracer/option.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TracerOption = void 0;
const tslib_1 = require("tslib");
const gConsts = tslib_1.__importStar(require("../consts"));
const gUtils = tslib_1.__importStar(require("../utils"));
const status = tslib_1.__importStar(require("../status"));
const configsClient = tslib_1.__importStar(require("../configsClient"));
const metrics = tslib_1.__importStar(require("../metrics"));
const propagator = tslib_1.__importStar(require("../propagator"));
const confProvider = tslib_1.__importStar(require("../confProvider"));
const breaker_1 = require("../breaker/breaker");
const reporter_1 = require("../reporter");
const sampler_1 = require("../sampler");
const idGenerator_1 = require("./idGenerator");
const env_1 = require("@byted-service/env");
const logger_1 = require("../logger");
const transaction_1 = require("../transaction");
const condition_1 = require("../sampler/condition");
const tracerId_1 = require("./tracerId");
const confProvider_1 = require("../confProvider");
const tracerConfig_1 = require("./tracerConfig");
const noopLogger = new logger_1.NoopLogger();
const UNKNOWN = 'unknown';
const MAX_UINT_32 = 2 ** 32 - 2;
class TracerOption {
    static mergeTracerOptions(tracer, options) {
        const opt = new TracerOption(tracer, options);
        if (!opt.highTraceIdGenerator) {
            let base;
            if (opt.env.ipv6) {
                base = gUtils.fnv32Or64(opt.env.ipv6, '32');
            }
            else {
                base = gUtils.fnv32Or64(opt.env.ipv4, '32');
            }
            const idGenerator = new idGenerator_1.AccumulateIdGenerator({
                maxId: MAX_UINT_32,
                delta: 1,
                randomNumber: () => {
                    return gUtils.rangeValue(0, Math.min(0, MAX_UINT_32));
                },
            });
            opt.highTraceIdGenerator = () => {
                return base + idGenerator.genId();
            };
        }
        if (!opt.postTraceIDGenerator) {
            opt.postTraceIDGenerator = (logId) => {
                if (!logId || logId.length <= 0) {
                    /* istanbul ignore next */
                    return new tracerId_1.TraceId('0', '0');
                }
                return gUtils.hashToTraceID(logId);
            };
        }
        opt.initEnv();
        return opt;
    }
    constructor(tracer, options) {
        // Enable Send Recv bytes metrics
        this.enableSendRecvBytesMetrics = true;
        this.registerCustomSpanMetricTags = new Set();
        this.registerClientSpanMetricTags = new Set();
        this.registerServerSpanMetricTags = new Set();
        this.tracer = tracer;
        this.mergeTracerOptions(options);
        const randNum = gUtils.nowInUs();
        this.randomNumber = () => {
            return Math.floor(Math.random() * randNum + 1);
        };
        this.transactionIdGenerator = new idGenerator_1.AccumulateIdGenerator({
            randomNumber: this.randomNumber,
            maxId: Number.MAX_SAFE_INTEGER,
            delta: 1,
        });
        this.spanIdGenerator = new idGenerator_1.AccumulateIdGenerator({
            randomNumber: this.randomNumber,
            maxId: Number.MAX_SAFE_INTEGER,
            delta: 1,
        });
    }
    /**
     * 设置默认参数
     * @param options
     */
    mergeTracerOptions(options) {
        options.env = gUtils.transPropertyToTargetNoCheck({
            dc: (0, env_1.getIDC)(),
            cluster: (0, env_1.getCluster)(),
            stage: (0, env_1.getStage)(),
            ipv4: (0, env_1.getIPv4)(),
            ipv6: (0, env_1.getIPv6)(),
            env: (0, env_1.getEnv)(),
            podName: (0, env_1.getPodName)(),
            language: gConsts.LANGUAGE,
            serviceName: (0, env_1.getPSM)(),
        }, options.env);
        this.env = Object.assign({}, options.env);
        this.serializationType = gConsts.DEFAULT_PROTO_TYPE;
        this.maxOperationForLimiter = gConsts.MAX_OPERATION_FOR_LIMITER;
        this.logger = options.logger || logger_1.defaultLogger;
        if (options.spanLogger) {
            this.spanLogger = options.spanLogger;
        }
        if (options.serverSpanLogger) {
            this.serverSpanLogger = options.serverSpanLogger;
        }
        if (options.clientSpanLogger) {
            this.clientSpanLogger = options.clientSpanLogger;
        }
        if (typeof options.disableDynamicConfig === 'boolean') {
            this.disableDynamicConfig = options.disableDynamicConfig;
        }
        if (typeof options.disableMetrics === 'boolean') {
            this.disableMetrics = options.disableMetrics;
        }
        if (options.metricClientOptions) {
            this.metricClientOptions = options.metricClientOptions;
        }
        if (typeof options.enableSendRecvBytesMetrics === 'boolean') {
            this.enableSendRecvBytesMetrics =
                options.enableSendRecvBytesMetrics;
        }
        if (typeof options.enableCollectTracerStatus === 'boolean') {
            this.enableCollectTracerStatus = options.enableCollectTracerStatus;
        }
        if (typeof options.highTraceIdGenerator === 'function') {
            this.highTraceIdGenerator = options.highTraceIdGenerator;
        }
        if (typeof options.postTraceIDGenerator === 'function') {
            this.postTraceIDGenerator = options.postTraceIDGenerator;
        }
        if (typeof options.samplingSeedGenerator === 'function') {
            this.samplingSeedGenerator = options.samplingSeedGenerator;
        }
        if (typeof options.disableReportRemoteAddrWhenErr === 'boolean') {
            this.disableReportRemoteAddrWhenErr =
                options.disableReportRemoteAddrWhenErr;
        }
        if (typeof options.disableReportSpan === 'boolean') {
            this.disableReportSpan = options.disableReportSpan;
        }
    }
    /**
     * 拉取 tcc 配置进行初始化
     */
    async syncInit() {
        await this.syncInitConfigurator();
        this.initMetricsClient();
        this.initSpanLogger();
        this.initStatusCounter();
        await this.syncInitBreaker();
        await this.syncInitSampler();
        this.initReporter();
        this.initPropagators();
        this.reportSdkVersion();
        this.registerDynamicConfig();
    }
    /**
     * 先使用默认配置，初始化 tracer，然后拉取 tcc 配置
     */
    asyncInit() {
        this.asyncInitConfigurator();
        this.initMetricsClient();
        this.initSpanLogger();
        this.initStatusCounter();
        this.asyncInitBreaker();
        this.asyncInitSampler();
        this.initReporter();
        this.initPropagators();
        this.reportSdkVersion();
        this.registerDynamicConfig();
    }
    reportSdkVersion() {
        const txn = new transaction_1.Transaction(this.tracer);
        txn.setTagKv(gConsts.SDK_VERSION_TAG, gConsts.SDK_VERSION);
        this.reporter.report(txn);
    }
    initPropagators() {
        const tp = new propagator.TextMapPropagator(gConsts.PropagatorTypes.TEXT_MAP, this.logger);
        const htp = new propagator.TextMapPropagator(gConsts.PropagatorTypes.HTTP_HEADER, this.logger);
        this.propagators = {
            [gConsts.TEXT_MAP]: tp,
            [gConsts.HTTP_HEADER]: htp,
        };
    }
    registerDynamicConfig() {
        const { configurator } = this;
        // 如果是静态配置器，注册后不会执行
        configurator.register(confProvider.NAMESPACE_CB, (config) => {
            return this.breaker.syncConfig(config);
        });
        configurator.register(confProvider.NAMESPACE_TRACER, _ => {
            return this.tracerConfig.syncConfig();
        });
        configurator.register(confProvider.NAMESPACE_SAMPLER, (config) => {
            this.sampler.syncConfig(config);
            return this.conditionSampler.syncConfig(config.conditionSamplingFilters);
        });
    }
    initReporter() {
        const reporterOpt = reporter_1.AgentReporter.getDefaultAgentReporterOption();
        reporterOpt.logger = this.logger;
        this.reporter = new reporter_1.AgentReporter(reporterOpt, this.env, this.statusCounter);
    }
    /**
     * 拉取配置后进行初始化
     */
    async syncInitSampler() {
        const samplerConfig = await this.configurator.pullSamplerConfig();
        this.sampler = new sampler_1.ByteSampler(samplerConfig, this.env, this.statusCounter);
        this.conditionSampler = new condition_1.ConditionSampler();
        this.conditionSampler.syncConfig(samplerConfig.conditionSamplingFilters);
    }
    /**
     * 先使用默认的配置，拉取 tcc 后使用 tcc 的配置
     */
    asyncInitSampler() {
        this.sampler = new sampler_1.ByteSampler(confProvider_1.DEFAULT_SAMPLER_CONFIG, this.env, this.statusCounter);
        this.conditionSampler = new condition_1.ConditionSampler();
        this.conditionSampler.syncConfig(confProvider_1.DEFAULT_SAMPLER_CONFIG.conditionSamplingFilters);
        this.configurator.pullSamplerConfig(true).then(samplerConfig => {
            // 如果返回为默认配置则不需要再次 init
            if (confProvider_1.DEFAULT_SAMPLER_CONFIG !== samplerConfig) {
                this.sampler.syncConfig(samplerConfig);
                this.conditionSampler.syncConfig(samplerConfig.conditionSamplingFilters);
            }
        });
    }
    /**
     * 先使用默认配置进行初始化
     */
    asyncInitBreaker() {
        const { tracer } = this;
        this.breaker = new breaker_1.Breaker(this.env, tracer, this.logger);
        this.breaker.asyncInit(this.configurator);
    }
    /**
     * 拉取 tcc 配置后进行初始化
     */
    async syncInitBreaker() {
        const { tracer } = this;
        this.breaker = new breaker_1.Breaker(this.env, tracer, this.logger);
        await this.breaker.syncInit(this.configurator);
    }
    initMetricsClient() {
        if (this.disableMetrics) {
            this.metricsClient = new metrics.NoopByteTraceMetrics();
        }
        else {
            const globalTags = {
                [gConsts.PSM_KEY]: this.env.serviceName || '-',
                [gConsts.DC_KEY]: this.env.dc || '-',
                [gConsts.CLUSTER_KEY]: this.env.cluster || '-',
                [gConsts.IPV4_KEY]: this.env.ipv4 || '-',
                [gConsts.IPV6_KEY]: this.env.ipv6 || '-',
                [gConsts.POD_NAME_KEY]: this.env.podName || '-',
                [gConsts.DEPLOY_STAGE_KEY]: this.env.stage || '-',
                [gConsts.ENV_KEY]: this.env.env || '-',
                [gConsts.FRAMEWOKER]: this.env.framework || 'gulu',
                [gConsts.TCE_PHYSICAL_CLUSTER_TAG]: process.env.TCE_PHYSICAL_CLUSTER || '-',
                [gConsts.TCE_NODE_LEVEL_TAG]: process.env.TCE_LOGICAL_CLUSTER || '-',
            };
            this.metricsClient = new metrics.ByteTraceMetrics(Object.assign({ defaultPrefix: `${gConsts.METRICS_TENANT}.${gConsts.METRICS_COMPONENT}` }, this.metricClientOptions), globalTags);
        }
    }
    initStatusCounter() {
        this.statusCounter = new status.StatusCounter(this.logger, this.metricsClient, this.enableCollectTracerStatus);
    }
    initSpanLogger() {
        if (this.spanLogger) {
            this.spanLogger = this.spanLogger;
        }
        else {
            this.spanLogger = noopLogger;
            this.logger.warn('use noopLogger as span logger.');
        }
        if (this.serverSpanLogger) {
            this.serverSpanLogger = this.serverSpanLogger;
        }
        else {
            this.serverSpanLogger = noopLogger;
            this.logger.warn('use noopLogger as server span logger.');
        }
        if (this.clientSpanLogger) {
            this.clientSpanLogger = this.clientSpanLogger;
        }
        else {
            this.clientSpanLogger = noopLogger;
            this.logger.warn('use noopLogger as client span logger.');
        }
    }
    /**
     * 先使用默认配置进行初始化，然后异步从 tcc 拉取配置
     */
    asyncInitConfigurator() {
        if (this.disableDynamicConfig) {
            this.configurator = new configsClient.StaticConfigurator();
        }
        else {
            this.configurator = configsClient.Configurator.create(this.env.serviceName, this.env.cluster, this.env.framework || 'default', this.logger);
            this.configurator.asyncInit();
        }
        this.tracerConfig = new tracerConfig_1.TracerConfig(this.configurator);
        this.tracerConfig.asyncConfig();
    }
    /**
     * 先拉取 tcc 配置再进行初始化
     */
    async syncInitConfigurator() {
        if (this.disableDynamicConfig) {
            this.configurator = new configsClient.StaticConfigurator();
        }
        else {
            this.configurator = configsClient.Configurator.create(this.env.serviceName, this.env.cluster, this.env.framework || 'default', this.logger);
            await this.configurator.syncInit();
        }
        this.tracerConfig = new tracerConfig_1.TracerConfig(this.configurator);
        await this.tracerConfig.syncConfig();
    }
    initEnv() {
        const { env } = this;
        if (env.ipv6) {
            env.ipv6Bytes = gUtils.ipv6ToBytes(env.ipv6);
        }
        else {
            env.ipv6Bytes = Buffer.from([]);
            env.ipv6 = UNKNOWN;
        }
        if (env.ipv4) {
            env.ipv4Uint32 = gUtils.ipv4ToInt(env.ipv4);
        }
        else {
            env.ipv4Uint32 = 0;
            env.ipv4 = UNKNOWN;
        }
    }
}
exports.TracerOption = TracerOption;
