// @byted-service/bytedtrace/lib/tracer/tracer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BytedTracer = void 0;
const tslib_1 = require("tslib");
const gTypes = tslib_1.__importStar(require("../types"));
const gConsts = tslib_1.__importStar(require("../consts"));
const gUtils = tslib_1.__importStar(require("../utils"));
const status = tslib_1.__importStar(require("../status"));
const kspan = tslib_1.__importStar(require("../span"));
const transaction_1 = require("../transaction");
const logger_1 = require("@byted-service/logger");
const event_1 = require("../event/event");
const option_1 = require("./option");
const tracerId_1 = require("./tracerId");
const env_1 = require("@byted-service/env");
const path_1 = tslib_1.__importDefault(require("path"));
const logger_2 = require("../logger");
const span_1 = require("../span");
const fnv_1 = require("@byted-service/fnv");
const defaultEventLogCallDepth = 4;
const eventMetricNamePrefix = 'event.';
const customMetricsNamePrefix = 'custom.';
// 全局只保持一个 BytedTracer 实例
const _global = global;
const defaultBytedTracer = Symbol.for('defaultBytedTracer');
class BytedTracer {
    /**
     * 注册全局实例
     * 此方法只能由框架(例如gulu)调用，中间件，lib，sdk 不能调用进行注册
     * @param tracer
     */
    static registerGlobalTracer(tracer) {
        const old = _global[defaultBytedTracer];
        if (old === tracer) {
            return;
        }
        if (old) {
            old.close();
            // 避免原型链 loop
            Object.setPrototypeOf(old, tracer);
            const keys = Object.getOwnPropertyNames(old);
            for (let i = 0; i < keys.length; i++) {
                delete old[keys[i]];
            }
        }
        _global[defaultBytedTracer] = tracer;
    }
    /**
     * 判断是否已经注册了全局实例
     * @returns
     */
    static isRegisteredGlobalTracer() {
        return !!_global[defaultBytedTracer];
    }
    /**
     * 拉取 tcc 配置后初始化 bytedtracer 实例
     * @param options
     * @returns
     */
    static async create(options) {
        const that = new BytedTracer(options);
        await that.options.syncInit();
        return that;
    }
    /**
     * 先使用默认配置，后续再拉取 tcc 配置
     * @param options
     * @returns
     */
    static createAsyncInit(options) {
        const that = new BytedTracer(options);
        that.options.asyncInit();
        return that;
    }
    /**
     * 获取一个全局唯一的实例，也是先使用默认配置后，后续拉取 tcc 配置
     * 如果全局实例没有将会返回一个全新的实例，并注册为全局实例！
     * @param options
     * @returns
     */
    static defaultBytedTracer(options = {}) {
        if (_global[defaultBytedTracer]) {
            return _global[defaultBytedTracer];
        }
        _global[defaultBytedTracer] = this.createDefaultBytedTracer(options);
        return _global[defaultBytedTracer];
    }
    /**
     * 创建默认 BytedTracer 实例
     * @param options
     * @returns
     */
    static createDefaultBytedTracer(options = {}) {
        const isDocker = (0, env_1.isInTCE)();
        options = gUtils.transPropertyToTargetNoCheck({
            enableCollectTracerStatus: true,
        }, options);
        if (!options.serverSpanLogger) {
            const accessLoggerOptions = {
                dir: isDocker
                    ? '/opt/tiger/toutiao/log/app'
                    : path_1.default.resolve(process.cwd(), './log/app'),
                level: 'trace',
                enableFileLog: true,
                enableAgentLog: false,
                ext: '.access.log',
            };
            options.serverSpanLogger = new logger_1.Logger(accessLoggerOptions);
        }
        if (!options.clientSpanLogger) {
            const callLoggerOptions = {
                dir: isDocker
                    ? '/opt/tiger/toutiao/log/rpc'
                    : path_1.default.resolve(process.cwd(), './log/rpc'),
                level: 'trace',
                enableFileLog: true,
                enableAgentLog: false,
                ext: '.call.log',
            };
            options.clientSpanLogger = new logger_1.Logger(callLoggerOptions);
        }
        if (!options.spanLogger) {
            options.spanLogger = options.clientSpanLogger;
        }
        options.env = gUtils.transPropertyToTargetNoCheck({
            serviceName: (0, env_1.getPSM)(),
            dc: (0, env_1.getIDC)(),
            stage: (0, env_1.getStage)(),
            cluster: (0, env_1.getCluster)(),
            env: (0, env_1.getEnv)(),
            podName: (0, env_1.getPodName)(),
            framework: 'default',
        }, options.env || {});
        return BytedTracer.createAsyncInit(options);
    }
    constructor(options) {
        this.options = option_1.TracerOption.mergeTracerOptions(this, options);
    }
    /**
     * 修改 server span logger
     * @param logger
     */
    setServerSpanLogger(logger) {
        this.options.serverSpanLogger = logger;
    }
    /**
     * 修改 client span logger 实例
     * @param logger
     */
    setClientSpanLogger(logger) {
        this.options.clientSpanLogger = logger;
    }
    /**
     * 修改 custom span logger 实例
     * @param logger
     */
    setCustomSpanLogger(logger) {
        this.options.spanLogger = logger;
    }
    /**
     * 创建 server 类型的 span
     * @param name span 名字
     * @param startOptions 其他选项
     * @returns
     */
    startServerSpan(name, startOptions = {}) {
        return this.startSpan(name, gTypes.SpanType.server, true, startOptions);
    }
    /**
     * 创建 client 类型的 span
     * @param name span 名字
     * @param startOptions 其他选项
     * @returns
     */
    startClientSpan(name, startOptions = {}) {
        const span = this.startSpan(name, gTypes.SpanType.client, true, startOptions);
        span.setFromService(name);
        return span;
    }
    /**
     * 创建自定义类型的 span
     * @param spanType span 类型
     * @param name span 名字
     * @param startOptions
     * @returns
     */
    startCustomSpan(spanType, name, startOptions = {}) {
        /* istanbul ignore if */
        if (spanType === gTypes.SpanType.server ||
            spanType === gTypes.SpanType.client) {
            throw new Error("[bytetrace] custom span type can not be 'server' or 'client'!");
        }
        try {
            return this.startSpan(name, spanType, true, startOptions);
        }
        catch (error) {
            this.options.statusCounter.recordWarningWithRateLimit(true, status.ErrPanic, "this panic won't affect your main function." +
                "It's a failure happend during monitoring data collection error: %o", error);
            return new span_1.NoopSpan(this);
        }
    }
    /**
     * 给 span 添加多个事件
     * @param events
     * @param span
     */
    addEvents(events, span) {
        if (this.options.tracerConfig.enableTrace() && events) {
            if (span) {
                events.forEach((event) => event.incCallDepthForLog(1));
                span.addEvents(events);
            }
            else {
                this._dealEvents(events, undefined);
            }
        }
    }
    /**
     * 给 span 添加单个事件
     * @param event
     * @param span
     */
    addEvent(event, span) {
        this.addEvents([event], span);
    }
    /**
     * metrics 打点 level
     * ServerSpan和ClientSpan metrics的级别
     * 0: 全量tag模式
     * 1: 精简tag模式
     * 2: 关闭模式
     * @returns
     */
    getSpanMetricsLevel() {
        return this.options.tracerConfig.getSpanMetricsLevel();
    }
    /**
     * 打 timer 类型的 metrics 点
     * @param metricsName
     * @param value
     * @param tagKvs
     * @returns
     */
    emitMetricsTimer(metricsName, value, tagKvs) {
        if (!this.options.tracerConfig.enableTrace()) {
            /* istanbul ignore next */
            return;
        }
        this.options.metricsClient.emitTimer(customMetricsNamePrefix + metricsName, value, undefined, tagKvs);
    }
    /**
     * 打 rate counter 类型的 metrics 点
     * @param metricsName
     * @param value
     * @param tagKvs
     * @returns
     */
    emitMetricsRateCounter(metricsName, value, tagKvs) {
        if (!this.options.tracerConfig.enableTrace()) {
            /* istanbul ignore next */
            return;
        }
        this.options.metricsClient.emitRateCounter(customMetricsNamePrefix + metricsName, value, undefined, tagKvs);
    }
    /**
     * 打 store 类型的 metrics 点
     * @param metricsName
     * @param value
     * @param tagKvs
     * @returns
     */
    emitMetricsStore(metricsName, value, tagKvs) {
        if (!this.options.tracerConfig.enableTrace()) {
            /* istanbul ignore next */
            return;
        }
        this.options.metricsClient.emitStore(customMetricsNamePrefix + metricsName, value, undefined, tagKvs);
    }
    /**
     * 打 counter 类型的 metrics 点
     * @param metricsName
     * @param value
     * @param tagKvs
     * @returns
     */
    emitMetricsCounter(metricsName, value, tagKvs) {
        if (!this.options.tracerConfig.enableTrace()) {
            /* istanbul ignore next */
            return;
        }
        this.options.metricsClient.emitCounter(customMetricsNamePrefix + metricsName, value, undefined, tagKvs);
    }
    /**
     * 打 meter 类型的 metrics 点
     * @param metricsName
     * @param value
     * @param tagKvs
     * @returns
     */
    emitMetricsMeter(metricsName, value, tagKvs) {
        if (!this.options.tracerConfig.enableTrace()) {
            /* istanbul ignore next */
            return;
        }
        this.options.metricsClient.emitMeter(customMetricsNamePrefix + metricsName, value, undefined, tagKvs);
    }
    /**
     * 关闭
     */
    close() {
        if (this.applyClose()) {
            this.options.sampler.close();
            this.options.reporter.close();
            this.options.statusCounter.close();
            this.options.configurator.close();
        }
    }
    /**
     * 从 carrier 中解析到 span context
     * @param format
     * @param carrier
     * @returns
     */
    bytedExtract(format, carrier) {
        const propagator = this.options.propagators[format];
        if (propagator) {
            return (propagator.extract(carrier) ||
                span_1.defaultNoopSpanContext);
        }
        return span_1.defaultNoopSpanContext;
    }
    /**
     * 将 span context 注入到 carrier 中
     * @param ctx span context
     * @param format propagator 类型
     * @param carrier 挂载对象
     * @returns
     */
    bytedInject(ctx, format, carrier) {
        const propagator = this.options.propagators[format];
        if (propagator) {
            propagator.inject(ctx, carrier);
        }
    }
    /**
     * 是否禁止了 postTrace
     * @returns
     */
    isPostTraceDisable() {
        return !this.options.tracerConfig.enablePostTrace();
    }
    /**
     * 是否禁止了染色采样
     * @returns
     */
    isDyeDisable() {
        return !this.options.tracerConfig.enableDyedHeader();
    }
    /**
     * 获取 service name
     * @returns
     */
    getServiceName() {
        return this.options.env.serviceName;
    }
    /**
     * 打印 span log
     * @param span
     */
    emitSpanLog(span) {
        try {
            if (span.isClientSpan()) {
                this.options.clientSpanLogger.log(span._buildTraceLogSpan());
            }
            else if (span.isServerSpan()) {
                this.options.serverSpanLogger.log(span._buildTraceLogSpan());
            }
            else {
                this.options.spanLogger.log(span._buildTraceLogByCustomSpan());
            }
        }
        catch (error) {
            this.options.statusCounter.recordWarningWithRateLimit(false, status.ErrLogSpanFail, 'log Span err:%o', error);
        }
    }
    /**
     * 创建 event
     * @param eventType
     * @param name
     * @param logLevel
     * @returns
     */
    newEvent(eventType, name, logLevel) {
        return new event_1.TraceEvent({
            tags: {},
            eventType,
            name,
            logLevel,
        });
    }
    /**
     * 是否是全量上报
     * @returns
     */
    allReported() {
        return this.options.tracerConfig.enableReportTotalSpan();
    }
    /**
     * 开启轻量级打点
     * @param component
     * @returns
     */
    enableThriftStyleMetrics(component) {
        return !this.options.tracerConfig.isThriftStyleMetricsDisable(component);
    }
    /**
     * 是否是轻量级模式
     * @param component
     * @returns
     */
    enableLightMode(component) {
        return this.options.tracerConfig.isLightModeEnable(component);
    }
    async _recordRootSampled(name) {
        try {
            await this.options.statusCounter.inc(name, 1, true);
        }
        catch (error) /* istanbul ignore next */ {
            this.options.statusCounter.recordWarningWithRateLimit(false, status.ErrIncRootSampledQpsCounterFail, 'inc root sampled qps fail error: %o', error);
        }
    }
    _genTraceId() {
        const high = this.options.highTraceIdGenerator();
        const low = this.randomId();
        return new tracerId_1.TraceId(high, low);
    }
    _dealEvents(events, span) {
        events.forEach((e) => {
            if (e instanceof event_1.TraceEvent) {
                if (e.getEmitLog()) {
                    this.emitLog(e, span);
                }
                if (e.getEmitMetrics()) {
                    this.emitMetricsForEvent(span, e);
                }
            }
        });
    }
    /**
     * 返回 sdk 内部状态信息
     * @returns
     */
    async getStatus() {
        const reportingQps = this.options.statusCounter.getRate(status.CounterTxnReportQps);
        const reportingBps = this.options.statusCounter.getRate(status.CounterTxnReportBps);
        const globalTracerConfig = await this.options.configurator.pullGlobalTracerConfig();
        const breakerConfig = await this.options.configurator.pullBreakerConfig();
        const samplerConfig = await this.options.configurator.pullSamplerConfig();
        return {
            tracerConfigs: {
                globalTracerConfig: JSON.stringify(globalTracerConfig),
                breakerConfig: JSON.stringify(breakerConfig),
                samplerConfig: JSON.stringify(samplerConfig),
            },
            reportingBps,
            reportingQps,
        };
    }
    appendSpanMetricTags(spanType, tagKey) {
        switch (spanType) {
            case gTypes.SpanType.client: {
                this.options.registerClientSpanMetricTags.add(tagKey);
                break;
            }
            case gTypes.SpanType.server: {
                this.options.registerServerSpanMetricTags.add(tagKey);
                break;
            }
            case gTypes.SpanType.custom: {
                this.options.registerCustomSpanMetricTags.add(tagKey);
                break;
            }
        }
    }
    samplingTransaction(seed, operationName, light) {
        const { isSampled, samplerType, samplingRate } = this.options.sampler.isSampled(seed, operationName);
        if (light && !isSampled) {
            return;
        }
        const txn = new transaction_1.Transaction(this);
        txn.setSamplingInfo(samplerType, samplingRate);
        logger_2.debugLogger.debug('samplingTransaction isSampled: ' + isSampled);
        if (isSampled) {
            if (txn.initSampled(operationName, samplerType === gConsts.LOWER_BOUND_SAMPLER)) {
                if (samplerType === gConsts.LOWER_BOUND_SAMPLER) {
                    this._recordRootSampled(status.ROOT_LOWER_BOUND_SAMPLED_QPS);
                }
                else {
                    this._recordRootSampled(status.ROOT_SAMPLED_QPS);
                }
            }
        }
        return txn;
    }
    enableSampling() {
        return this.options.tracerConfig.enableSampling();
    }
    genSpanId() {
        return this.options.spanIdGenerator.genId() + '';
    }
    /**
     * @param ctx
     * @param operationName
     * @param parentSpanReference
     * @param enableSampling
     * @returns
     */
    newTransaction(options) {
        let transaction;
        const { logId, operationName, parentSpanReference, enableSampling, light, } = options;
        // 如果使用主动采样
        if (enableSampling && this.enableSampling()) {
            if (!parentSpanReference ||
                ((!parentSpanReference.referencedContext.traceID ||
                    !parentSpanReference.referencedContext.traceID.isValid()) &&
                    !parentSpanReference.referencedContext.isDebug())) {
                logger_2.debugLogger.debug(`[bytetrace] newTransaction parentSpanReference is ${parentSpanReference}`);
                // 如果没有自定义采样种子,使用默认的采样算法
                if (typeof this.options.samplingSeedGenerator !== 'function') {
                    if (logId) {
                        // 使用logID哈希作为采样种子
                        transaction = this.samplingTransaction((0, fnv_1.logIdHash)(logId), operationName, light);
                    }
                    else {
                        // 不执行采样,但是要记录qps
                        this.options.sampler.incAndGetSamplingRate(operationName);
                    }
                }
                else {
                    // 使用自定义的采样种子
                    transaction = this.samplingTransaction(this.options.samplingSeedGenerator(), operationName, light);
                }
            }
            else {
                // 父context中采样信息不为空,不需要主动采样,但是要记录qps
                let samplingRate;
                try {
                    samplingRate =
                        this.options.sampler.incAndGetSamplingRate(operationName);
                }
                catch (error) {
                    /* istanbul ignore next */
                    this.options.statusCounter.recordWarningWithRateLimit(false, status.ErrIncAndGetSamplingRateFail, 'incAndGetSamplingRate failed,error: %o', error);
                }
                if (light &&
                    !gUtils.isSampledContext(parentSpanReference.referencedContext)) {
                    /* istanbul ignore next */
                    return;
                }
                transaction = new transaction_1.Transaction(this);
                transaction.initRootSpanContext(parentSpanReference.referencedContext, samplingRate, operationName, true);
            }
        }
        else if (parentSpanReference) {
            if (light &&
                !gUtils.isSampledContext(parentSpanReference.referencedContext)) {
                return;
            }
            transaction = new transaction_1.Transaction(this);
            transaction.initRootSpanContext(parentSpanReference.referencedContext, -1, operationName, true);
        }
        if (!transaction) {
            if (light) {
                return;
            }
            transaction = new transaction_1.Transaction(this);
        }
        if (!transaction.traceID.isValid()) {
            if (this.allReported() || transaction.isSampled()) {
                transaction.traceID = this._genTraceId();
            }
        }
        transaction.transactionId = this.options.transactionIdGenerator.genId();
        transaction.logId = logId;
        return transaction;
    }
    randomId() {
        return this.options.randomNumber();
    }
    startSpan(name, spanType, enableSampling, options = {}) {
        const { tracerConfig } = this.options;
        if (!tracerConfig.enableTrace()) {
            return new span_1.NoopSpan(this);
        }
        let light = false;
        let txn;
        let sp;
        let spanRefType = 0;
        let parentSpanId = '0';
        let baggage;
        let txnFinished = false;
        if (options && options.lightModeCheckComponent) {
            light = tracerConfig.isLightModeEnable(options.lightModeCheckComponent);
            if (light) {
                options.emitMetrics = false;
                options.emitSpanLog = false;
            }
        }
        // 找父 span
        const parentSpanReference = this.findParent(options || {});
        // 如果传入的 parentSpanReference 不是 spanContext 实例直接返回 noopSpan
        if (parentSpanReference && parentSpanReference.referencedContext) {
            if (!(parentSpanReference.referencedContext instanceof
                kspan.SpanContext) ||
                parentSpanReference.referencedContext.noop()) {
                return new span_1.NoopSpan(this);
            }
            spanRefType = parentSpanReference.referenceType;
            parentSpanId = parentSpanReference.referencedContext.spanID;
        }
        else {
            // 进程内不需要开启新的 txn 与父亲 span 共用一个 txn
            const parentSpan = options === null || options === void 0 ? void 0 : options.parentSpan;
            if (parentSpan) {
                spanRefType = gConsts.CHILD_OF_REF;
                parentSpanId = parentSpan.spanId;
                baggage = parentSpan.getBaggage();
                txn = parentSpan.transaction;
                if (txn) {
                    if (txn.isFinished()) {
                        parentSpanId = '0';
                        baggage = undefined;
                        txnFinished = true;
                    }
                    else {
                        if (light && !txn.isSampled()) {
                            return new span_1.NoopSpan(this);
                        }
                    }
                }
            }
        }
        const isRootSpan = !txn || txnFinished;
        sp = new kspan.Span(this);
        sp._resetReservedMetricsParam();
        sp.spanId = this.genSpanId();
        if (isRootSpan) {
            txn = this.newTransaction({
                logId: options.logId,
                operationName: name,
                parentSpanReference,
                enableSampling,
                light,
            });
            if (!txn) {
                return new span_1.NoopSpan(this);
            }
            txn.setRootSpan(sp);
            txn.method = name;
        }
        sp.parentSpanId = parentSpanId;
        sp._init({
            name,
            spanType,
            spanRefType,
            txn: txn,
            isRootSpan,
            baggage,
        });
        sp._synStartOpts(options);
        sp.setConditionFilters(this.options.conditionSampler.getFilters(sp.getSpanType(), sp.getName()));
        return sp;
    }
    findParent(options) {
        if (!options || !options.reference || options.reference.length === 0) {
            return;
        }
        return options.reference[options.reference.length - 1];
    }
    buildEventData(event) {
        const datas = [];
        datas.push('_type:');
        datas.push(event.getEventType());
        datas.push(' _name:');
        datas.push(event.getName());
        const tags = event.getTags();
        let len = Object.keys(tags).length;
        if (len > 0) {
            datas.push(' _tags:');
            const keys = Object.keys(tags);
            for (let i = 0; i < keys.length; i++) {
                const key = keys[i];
                const tag = tags[key];
                datas.push(key);
                datas.push('=');
                datas.push(gUtils.stringify(tag));
                if (--len > 0) {
                    datas.push(',');
                }
            }
        }
        datas.push(' _msg:');
        datas.push(event.getContent());
        return datas.join('');
    }
    emitLog(event, sp) {
        const logData = {};
        logData.location = logger_1.Logger.getStackFilePath(event.getCallDepth() + defaultEventLogCallDepth);
        logData.message = this.buildEventData(event);
        if (sp) {
            logData.tags = {};
            logData.tags[gConsts.LOG_ID_KEY] = sp.getLogId();
            logData.tags[gConsts.SPAN_IDKEY] = sp.spanId;
        }
        const LogLevels = gConsts.LogLevels;
        switch (event.getLogLevel()) {
            case LogLevels.LEVEL_TRACE: {
                logData.level = 'trace';
                break;
            }
            case LogLevels.LEVEL_DEBUG: {
                logData.level = 'debug';
                break;
            }
            case LogLevels.LEVEL_INFO: {
                logData.level = 'info';
                break;
            }
            case LogLevels.LEVEL_WARN: {
                logData.level = 'warn';
                break;
            }
            case LogLevels.LEVEL_ERROR: {
                logData.level = 'error';
                break;
            }
            case LogLevels.LEVEL_FATAL: {
                logData.level = 'fatal';
                break;
            }
            default: {
                logData.level = 'fatal';
            }
        }
        this.options.logger.log(logData);
    }
    emitMetricsForEvent(sp, event) {
        if (!this.options.tracerConfig.isMetricsEnable(true)) {
            return;
        }
        let method = '-';
        if (sp) {
            method = sp.getRootSpanMethod();
        }
        const tags = {
            [gConsts.NAME_KEY]: event.getName(),
            [gConsts.METHOD_KEY]: method,
        };
        const isLevelNoDefine = event.getLogLevel() !== gConsts.LogLevels.LEVEL_NO_DEFINE;
        const eventTags = event.getTags();
        const keys = Object.keys(eventTags);
        for (let i = 0; i < keys.length; i++) {
            tags[keys[i]] = isLevelNoDefine ? eventTags[keys[i]] : '-';
        }
        this.options.metricsClient.emitRateCounter(eventMetricNamePrefix + event.getEventType(), 1, undefined, tags);
    }
    applyClose() {
        if (!this.isClose) {
            this.isClose = true;
            return true;
        }
        /* istanbul ignore next */
        return false;
    }
}
exports.BytedTracer = BytedTracer;
