// @universe-design/react/src/tabs/stories/homeless.stories.tsx
import React, { useState } from 'react';
import type { DefaultSizeType, TabsType } from '../..';
import { Tabs, Radio, Space, Divider, Slider, Button, Illustration, Checkbox } from '../..';
import './index.less';
import { EmptyNegativeError } from '@universe-design/illustration-react';
import ContactsOutlined from '@universe-design/icons-react/ContactsOutlined';

export default {
  title: 'Components/Tabs/临时示例',
  component: Tabs,
};

const { TabPane } = Tabs;

// let count = 5;

export const AllTabs = () => (
  <div>
    <Divider text="基础" />
    <div style={{ padding: 10 }}>
      <Tabs>
        <TabPane tab="tab1">
          <div>TabPane content1</div>
        </TabPane>
        <TabPane tab="tab2" disabled>
          <div>
            <p>TabPane content2.</p>
            <p>TabPane content2.</p>
            <p>TabPane content2.</p>
            <p>TabPane content2.</p>
          </div>
        </TabPane>
        <TabPane tab="tab3">
          <div>TabPane content3</div>
        </TabPane>
      </Tabs>
    </div>

    <Divider text="卡片类型" />
    <div style={{ padding: 10 }}>
      <Tabs type="card">
        <TabPane tab="tab1">
          <div>TabPane content1</div>
        </TabPane>
        <TabPane tab="tab2" disabled>
          <div>
            <p>TabPane content2.</p>
            <p>TabPane content2.</p>
            <p>TabPane content2.</p>
            <p>TabPane content2.</p>
          </div>
        </TabPane>
        <TabPane tab="tab3">
          <div>TabPane content3</div>
        </TabPane>
        <TabPane tab="tab4">
          <div>TabPane content4</div>
        </TabPane>
        <TabPane tab="tab5">
          <div>TabPane content5</div>
        </TabPane>
      </Tabs>
    </div>

    <Divider text="自定义 tabs 间距" />
    <div style={{ padding: 10 }}>
      <Tabs tabBarGutter={10} scrollable scrollButtonGutter={8}>
        <TabPane tab="tab1">
          <div>TabPane content1</div>
        </TabPane>
        <TabPane tab="tab2">
          <div>TabPane content2</div>
        </TabPane>
        <TabPane tab="tab3">
          <div>TabPane content3</div>
        </TabPane>
        <TabPane tab="tab4">
          <div>TabPane content4</div>
        </TabPane>
        <TabPane tab="tab5">
          <div>TabPane content5</div>
        </TabPane>
        <TabPane tab="tab6">
          <div>TabPane content6</div>
        </TabPane>
      </Tabs>
    </div>

    <Divider text="展示 更多" />
    <div style={{ padding: 10 }}>
      <Tabs>
        {[...Array.from({ length: 20 }, (v, i) => i)].map((i, index) => (
          <TabPane key={String(index)} tab={`标签 ${index + 1}`} disabled={index === 18}>
            <div>TabPane content {index + 1} </div>
          </TabPane>
        ))}
      </Tabs>
    </div>
  </div>
);

export const ActiveKey = () => {
  const [activeKey, setActiveKey] = useState<string>('tab2');
  return (
    <div>
      <Radio.Group value={activeKey} onChange={v => setActiveKey(v as string)}>
        <Radio.Button value="tab1">tab1</Radio.Button>
        <Radio.Button value="tab2">tab2</Radio.Button>
        <Radio.Button value="tab3">tab3</Radio.Button>
      </Radio.Group>
      <div style={{ padding: 10 }}>
        <Tabs activeKey={activeKey}>
          <TabPane tab="tab1" key="tab1">
            <div>TabPane content1</div>
          </TabPane>
          <TabPane tab="tab2" key="tab2">
            <div>
              <p>TabPane content2.</p>
              <p>TabPane content2.</p>
              <p>TabPane content2.</p>
              <p>TabPane content2.</p>
            </div>
          </TabPane>
          <TabPane tab="tab3" key="tab3">
            <div>TabPane content3</div>
          </TabPane>
        </Tabs>
      </div>
    </div>
  );
};

export const Event = () => {
  const onChange = (key: string) => console.log('onChange, key is', key);
  return (
    <div style={{ padding: 10 }}>
      <Tabs onChange={onChange}>
        <TabPane tab="tab1" key="tab1">
          <div>TabPane content1</div>
        </TabPane>
        <TabPane tab="tab2" key="tab2">
          <div>
            <p>TabPane content2.</p>
            <p>TabPane content2.</p>
            <p>TabPane content2.</p>
            <p>TabPane content2.</p>
          </div>
        </TabPane>
        <TabPane tab="tab3" key="tab3">
          <div>TabPane content3</div>
        </TabPane>
      </Tabs>
    </div>
  );
};

// export const EditTab = () => {
//   const initTabs = [...new Array(count)].map((x, i) => ({
//     title: `Tab ${i + 1}`,
//     key: `key${i + 1}`,
//     content: `${i + 1}`,
//   }));
//   const [tabs, setTabs] = useState(initTabs);
//   const [activeTab, setActiveTab] = useState('key2');

//   const handleAddTab = () => {
//     count += 1;
//     const newTab = {
//       title: `New Tab${++count}`,
//       key: `new key${count}`,
//       content: `${count}`,
//     };
//     setTabs([...tabs, newTab]);
//     setActiveTab(newTab.key);
//   };

//   const handleDeleteTab = (key: string) => {
//     const index = tabs.findIndex(x => x.key === key);
//     const newTabs = tabs.slice(0, index).concat(tabs.slice(index + 1));

//     if (key === activeTab && index > -1 && newTabs.length) {
//       setActiveTab(
//         newTabs[index] ? newTabs[index].key : newTabs[index - 1].key
//       );
//     }

//     if (index > -1) {
//       setTabs(newTabs);
//     }
//   };

//   const handleOnEdit: onEditFunction = (type, key) => {
//     if (type === 'add') {
//       handleAddTab();
//     } else {
//       key && handleDeleteTab(key);
//     }
//   };

//   return (
//     <div>
//       <Divider text="卡片类型-可编辑" />
//       <Tabs activeKey={activeTab} type="editable-card" onEdit={handleOnEdit}>
//         {tabs.map((x, i) => (
//           <TabPane key={x.key} tab={x.title}>
//             <div>{`Content of Tab Panel ${x.content}`}</div>
//           </TabPane>
//         ))}
//       </Tabs>
//     </div>
//   );
// };

export { default as WidthBadge } from '../demo/4.with-badge';
export { default as Case0 } from '../demo/999.case0';
export { default as Case1 } from '../demo/999.case1';

export const ShowMore = () => {
  const [w, setW] = useState(400);
  const [size, setSize] = useState<DefaultSizeType>('medium');
  const [type, setType] = useState<TabsType>('line');

  return (
    <Space style={{ width: '100%' }} direction="vertical" size="large">
      <div style={{ width: '50%' }}>
        <span>宽度:</span>
        <Slider value={w} onChange={setW} step={100} min={400} max={1000} />
      </div>
      <Space size="large" align="center">
        <div>
          <span style={{ marginRight: 8 }}>Size: </span>
          <Radio.Group value={size} onChange={(v: any) => setSize(v)}>
            <Radio.Button value="large">Large</Radio.Button>
            <Radio.Button value="medium">Medium</Radio.Button>
            <Radio.Button value="small">Small</Radio.Button>
          </Radio.Group>
        </div>
        <div>
          <span style={{ marginRight: 8 }}>Type: </span>
          <Radio.Group value={type} onChange={(v: any) => setType(v)}>
            <Radio.Button value="line">Line</Radio.Button>
            <Radio.Button value="card">Card</Radio.Button>
            <Radio.Button value="tag">Tag</Radio.Button>
          </Radio.Group>
        </div>
      </Space>
      <Tabs style={{ width: w }} size={size} type={type}>
        {[...Array.from({ length: 15 }, (v, i) => i)].map((i, index) => (
          <TabPane key={String(index)} tab={`标签 ${index + 1}`} disabled={index === 12}>
            <div>TabPane content {index + 1} </div>
          </TabPane>
        ))}
      </Tabs>
    </Space>
  );
};

export const MaxCount = () => {
  const [count, setCount] = useState(8);
  const [size, setSize] = useState<DefaultSizeType>('medium');
  const [type, setType] = useState<TabsType>('line');

  return (
    <Space style={{ width: '100%' }} direction="vertical" size="large">
      <div style={{ width: '50%' }}>
        <span>最大展示数量:</span>
        <Slider value={count} onChange={setCount} step={1} min={5} max={15} />
      </div>
      <Space size="large" align="center">
        <div>
          <span style={{ marginRight: 8 }}>Size: </span>
          <Radio.Group value={size} onChange={(v: any) => setSize(v)}>
            <Radio.Button value="large">Large</Radio.Button>
            <Radio.Button value="medium">Medium</Radio.Button>
            <Radio.Button value="small">Small</Radio.Button>
          </Radio.Group>
        </div>
        <div>
          <span style={{ marginRight: 8 }}>Type: </span>
          <Radio.Group value={type} onChange={(v: any) => setType(v)}>
            <Radio.Button value="line">Line</Radio.Button>
            <Radio.Button value="card">Card</Radio.Button>
            <Radio.Button value="tag">Tag</Radio.Button>
          </Radio.Group>
        </div>
      </Space>
      <Tabs maxCount={count} size={size} type={type}>
        {[...Array.from({ length: 15 }, (v, i) => i)].map((i, index) => (
          <TabPane key={String(index)} tab={`标签 ${index + 1}`} disabled={index === 12}>
            <div>TabPane content {index + 1} </div>
          </TabPane>
        ))}
      </Tabs>
    </Space>
  );
};

export const TabBarPadding = () => (
  <div className="ud-demo-tabs-tab-bar-padding">
    <Tabs>
      <TabPane tab="tab1" key="tab1">
        <div>TabPane content1</div>
      </TabPane>
      <TabPane tab="tab2" key="tab2">
        <div>
          <p>TabPane content2.</p>
          <p>TabPane content2.</p>
          <p>TabPane content2.</p>
          <p>TabPane content2.</p>
        </div>
      </TabPane>
      <TabPane tab="tab3" key="tab3">
        <div>TabPane content3</div>
      </TabPane>
    </Tabs>
  </div>
);

export const TabBarExtraContent = () => (
  <>
    <Tabs tabBarExtraContent={<Button>Extra Action</Button>}>
      <TabPane tab="tab1" key="tab1">
        <div>TabPane content1</div>
      </TabPane>
      <TabPane tab="tab2" key="tab2">
        <div>
          <p>TabPane content2.</p>
          <p>TabPane content2.</p>
          <p>TabPane content2.</p>
          <p>TabPane content2.</p>
        </div>
      </TabPane>
      <TabPane tab="tab3" key="tab3">
        <div>TabPane content3</div>
      </TabPane>
    </Tabs>
    <Tabs
      tabBarExtraContent={{
        left: <Button>Extra Action</Button>,
      }}
    >
      <TabPane tab="tab1" key="tab1">
        <div>TabPane content1</div>
      </TabPane>
      <TabPane tab="tab2" key="tab2">
        <div>
          <p>TabPane content2.</p>
          <p>TabPane content2.</p>
          <p>TabPane content2.</p>
          <p>TabPane content2.</p>
        </div>
      </TabPane>
      <TabPane tab="tab3" key="tab3">
        <div>TabPane content3</div>
      </TabPane>
    </Tabs>
    <Tabs
      tabBarExtraContent={{
        left: <Button>Extra Action</Button>,
        right: <Button>Extra Action</Button>,
      }}
    >
      <TabPane tab="tab1" key="tab1">
        <div>TabPane content1</div>
      </TabPane>
      <TabPane tab="tab2" key="tab2">
        <div>
          <p>TabPane content2.</p>
          <p>TabPane content2.</p>
          <p>TabPane content2.</p>
          <p>TabPane content2.</p>
        </div>
      </TabPane>
      <TabPane tab="tab3" key="tab3">
        <div>TabPane content3</div>
      </TabPane>
    </Tabs>
  </>
);

export const DoubleClick = () => (
  <Tabs onTabDoubleClick={() => alert('Double clicked')}>
    <TabPane tab="tab1">
      <div>TabPane content1</div>
    </TabPane>
    <TabPane tab="tab2" disabled>
      <div>
        <p>TabPane content2.</p>
        <p>TabPane content2.</p>
        <p>TabPane content2.</p>
        <p>TabPane content2.</p>
      </div>
    </TabPane>
    <TabPane tab="tab3">
      <div>TabPane content3</div>
    </TabPane>
  </Tabs>
);

export const Stretch = () => {
  const paneStyle = {
    height: '100%',
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
  };
  return (
    <Tabs style={{ height: 400 }}>
      <TabPane tab="tab1" style={paneStyle}>
        <Illustration image={<EmptyNegativeError />} description="Error" />
      </TabPane>
    </Tabs>
  );
};

export const Scrollable = () => {
  const TAB_COUNT = 20;
  const [scrollable, setScrollable] = React.useState(true);
  const [activeKey, setActiveKey] = React.useState('tab1');
  return (
    <div style={{ maxWidth: 500 }}>
      <Checkbox checked={scrollable} onCheckChange={v => setScrollable(v)}>
        Tabs Scrollable
      </Checkbox>
      <Button onClick={() => setActiveKey('tab9')}>切换到 Tab9</Button>
      <Tabs scrollable={scrollable} onChange={key => setActiveKey(key)} activeKey={activeKey}>
        {new Array(TAB_COUNT).fill('').map((_, index) => (
          <TabPane
            key={`tab${index}`}
            tab={
              <span>
                <ContactsOutlined /> Tab {index}
              </span>
            }
          >
            Content of Tab Panel {index}
          </TabPane>
        ))}
      </Tabs>
    </div>
  );
};
