// @universe-design/react/src/tabs/stories/screenshot.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C999Case0 from '../demo/999.case0';
import C999Case1 from '../demo/999.case1';
import C999Scrollable from '../demo/999.scrollable';
import CNestedScreenshot from '../demo/nested.screenshot';
import CSizingBehavior from '../demo/sizing-behavior';
if (!(C999Case0 as Story).storyName) {
  (C999Case0 as Story).storyName = '999 case0';
}
if (!(C999Case1 as Story).storyName) {
  (C999Case1 as Story).storyName = '999 case1';
}
if (!(C999Scrollable as Story).storyName) {
  (C999Scrollable as Story).storyName = '999 scrollable';
}
if (!(CNestedScreenshot as Story).storyName) {
  (CNestedScreenshot as Story).storyName = 'Nested screenshot';
}
if (!(CSizingBehavior as Story).storyName) {
  (CSizingBehavior as Story).storyName = 'Sizing behavior';
}

export default {
  title: 'Components/Tabs/截图测试',
  tags: ['autodocs'],
};

export { C999Case0, C999Case1, C999Scrollable, CNestedScreenshot, CSizingBehavior };
