// @universe-design/react/src/tabs/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C4WithBadge from '../demo/4.with-badge';
import CBasic from '../demo/basic';
import CCard from '../demo/card';
import CController from '../demo/controller';
import CIcon from '../demo/icon';
import CPriority from '../demo/priority';
import CScrollable from '../demo/scrollable';
import CSize from '../demo/size';
import CSticky from '../demo/sticky';
import CTabsSizingBehavior from '../demo/tabs-sizing-behavior';
if (!(C4WithBadge as Story).storyName) {
  (C4WithBadge as Story).storyName = '5. 自定义「更多」页签';
  (C4WithBadge as Story).parameters = {
    docs: {
      description: {
        story:
          '使用 `tabMore` 插槽属性配合 `TabMore` `TabMoreContent` `TabMoreArrow` 组件开发更灵活的「更多」页签。\n\n注意，当页签内容过多时，需要配合 `TabContent` 做到内容单行收起，内容的默认最大宽度为 `240px`，此时如果还需配合 `Badge` 使用，要保证 `Badge` 包裹在 `TabContent` 外，以免徽标被裁剪。\n\n',
      },
    },
  };
}
if (!(CBasic as Story).storyName) {
  (CBasic as Story).storyName = '0. 基础用法';
  (CBasic as Story).parameters = {
    docs: {
      description: {
        story: '简单的使用。\n\n',
      },
    },
  };
}
if (!(CCard as Story).storyName) {
  (CCard as Story).storyName = '3. 卡片类型';
  (CCard as Story).parameters = {
    docs: {
      description: {
        story: '设置`type`为`card`可以开启卡片模式。\n\n',
      },
    },
  };
}
if (!(CController as Story).storyName) {
  (CController as Story).storyName = '1. 受控模式';
  (CController as Story).parameters = {
    docs: {
      description: {
        story: '通过`activeTab`开启受控模式。\n\n',
      },
    },
  };
}
if (!(CIcon as Story).storyName) {
  (CIcon as Story).storyName = '2. 带图标的标签';
  (CIcon as Story).parameters = {
    docs: {
      description: {
        story: '通过自定义的`tab`, 可以给标签加`icon`。\n\n',
      },
    },
  };
}
if (!(CPriority as Story).storyName) {
  (CPriority as Story).storyName = '7. 主从标签';
  (CPriority as Story).parameters = {
    docs: {
      description: {
        story: '自定义主从关系的 `TabsItem` 标签样式。\n\n',
      },
    },
  };
}
if (!(CScrollable as Story).storyName) {
  (CScrollable as Story).storyName = '6. 横向滚动';
  (CScrollable as Story).parameters = {
    docs: {
      description: {
        story: '通过`scrollable` 属性，让选项过多时水平滚动，而不是收起到[更多]。\n\n',
      },
    },
  };
}
if (!(CSize as Story).storyName) {
  (CSize as Story).storyName = '4. 三种大小';
  (CSize as Story).parameters = {
    docs: {
      description: {
        story: '标签页支持 3 种大小，分别为`large`、`medium`、`small`，默认大小为`medium`。\n\n',
      },
    },
  };
}
if (!(CSticky as Story).storyName) {
  (CSticky as Story).storyName = '7. sticky';
  (CSticky as Story).parameters = {
    docs: {
      description: {
        story: '通过`sticky` 属性，可以让 TabBar 固定在顶部。\n\n',
      },
    },
  };
}
if (!(CTabsSizingBehavior as Story).storyName) {
  (CTabsSizingBehavior as Story).storyName = '8. 布局方式';
  (CTabsSizingBehavior as Story).parameters = {
    docs: {
      description: {
        story: '通过 `tabsSizingBehavior` 属性，可以设置标签页的布局方式。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Tabs/官网示例',
  tags: ['autodocs'],
};

export { C4WithBadge, CBasic, CCard, CController, CIcon, CPriority, CScrollable, CSize, CSticky, CTabsSizingBehavior };
