// @byted-service/fetch-protocol-consul/lib/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.handle = exports.protocol = void 0;
const tslib_1 = require("tslib");
const consul_1 = require("@byted-service/consul");
const logger_1 = require("@byted-service/logger");
const node_fetch_1 = tslib_1.__importDefault(require("@byted/node-fetch"));
const consts_1 = require("./consts");
tslib_1.__exportStar(require("./typings"), exports);
let socketPath = '';
let port = 0;
if (((_a = process.env.TCE_ENABLE_HTTP_SIDECAR_EGRESS) === null || _a === void 0 ? void 0 : _a.toLowerCase()) === 'true') {
    if (((_b = process.env.WITH_HTTP_MESH_EGRESS_UDS) === null || _b === void 0 ? void 0 : _b.toLowerCase()) === 'true') {
        socketPath = process.env.SERVICE_MESH_HTTP_EGRESS_ADDR || '';
    }
    if (!socketPath) {
        port = Number(process.env.MESH_EGRESS_PORT);
    }
}
const inMesh = Boolean(socketPath || port);
const defaultConsul = new consul_1.Consul();
const logger = new logger_1.Logger();
const filterEndpointsByEnv = (endpoints, env) => {
    const result = [];
    for (let i = 0; i < endpoints.length; i++) {
        const endpoint = endpoints[i];
        const endpointEnv = endpoint.env || endpoint.tags.env; // 兼容 @byted-service/consul@1.x
        if (endpointEnv === env) {
            result.push(endpoint);
            continue;
        }
        if (env === 'prod' && (endpointEnv === 'canary' || !endpointEnv)) {
            result.push(endpoint);
        }
    }
    return result;
};
const weightedRandom = (endpoints) => {
    let total = 0;
    for (let i = 0; i < endpoints.length; i++) {
        total += endpoints[i].weight || Number(endpoints[i].tags.weight) || 50; // 兼容 @byted-service/consul@1.x
    }
    let rand = Math.floor(total * Math.random());
    for (let i = 0; i < endpoints.length; i++) {
        rand -= endpoints[i].weight || Number(endpoints[i].tags.weight) || 50; // 兼容 @byted-service/consul@1.x
        if (rand < 0) {
            return endpoints[i];
        }
    }
    /* istanbul ignore next */
    throw new Error('should never happen');
};
exports.protocol = 'consul';
const handle = async (ctx) => {
    const options = ctx.options;
    let consul = options.consul;
    let consulOptions = options.consulOptions;
    let consulLookupOptions = options.consulLookupOptions;
    if (consul && !(consul instanceof consul_1.Consul)) {
        consts_1.warnings.options.consul();
        consulLookupOptions = consulLookupOptions || consul;
        consul = undefined;
    }
    if (options.consulClient) {
        consts_1.warnings.options.consulClient();
        consul = consul || options.consulClient;
    }
    if (options.consulClientOptions) {
        consts_1.warnings.options.consulClientOptions();
        consulOptions = consulOptions || options.consulClientOptions;
    }
    if (!consul) {
        consul = consulOptions ? new consul_1.Consul(consulOptions) : defaultConsul;
    }
    const psm = (ctx.to = ctx.to || ctx.url.host);
    const cluster = (ctx.toCluster =
        (consulLookupOptions === null || consulLookupOptions === void 0 ? void 0 : consulLookupOptions.cluster) || ctx.toCluster || 'default');
    const idc = (ctx.toIdc = (consulLookupOptions === null || consulLookupOptions === void 0 ? void 0 : consulLookupOptions.idc) || ctx.toIdc || '');
    const env = (ctx.env = (consulLookupOptions === null || consulLookupOptions === void 0 ? void 0 : consulLookupOptions.env) || ctx.env || '');
    const mesh = inMesh && options.enableMesh !== false && !options.disableMesh;
    let resp = null;
    if (mesh) {
        ctx.mesh = '1';
        ctx.headers.set(consts_1.httpHeaders.HeaderDesService, psm);
        if (cluster) {
            ctx.headers.set(consts_1.httpHeaders.HeaderDesCluster, cluster);
        }
        if (idc) {
            ctx.headers.set(consts_1.httpHeaders.HeaderDesIDC, idc);
        }
        if (env) {
            ctx.headers.set(consts_1.httpHeaders.HeaderSourceEnv, env);
        }
        // 当用户主动传了 timeout 才传给 mesh
        if (typeof options.timeout === 'number') {
            ctx.headers.set(consts_1.httpHeaders.HeaderDesRequestTimeout, String(options.timeout || consts_1.meshMaxTimeout));
        }
        if (socketPath) {
            ctx.rip = socketPath;
            resp = await (0, node_fetch_1.default)(ctx.url, Object.assign(Object.assign({}, ctx.getFetchOptions()), { socketPath }));
        }
        else if (port) {
            ctx.rip = ctx.url.host = `127.0.0.1:${port}`;
            resp = await ctx.fetch();
        }
        if (resp === null || resp === void 0 ? void 0 : resp.headers) {
            // 取出的 ipv6 本身已经是 [ip]:port 形式，不用做任何处理
            ctx.rip = resp.headers.get('rip') || '';
        }
    }
    if (!resp) {
        ctx.mesh = '0';
        const endpoints = await (0, consul_1.lookupName)(psm, Object.assign({ cluster,
            idc, timeout: ctx.timer.remain('ms') }, consulLookupOptions), consul);
        let eps = [];
        if (env) {
            eps = filterEndpointsByEnv(endpoints, env);
        }
        // If get none instances, fallback to prod env
        if (eps.length === 0) {
            if (env) {
                logger.log({
                    level: 'warn',
                    logId: ctx.logId,
                    message: '[fetch] Can not find env "%s" for %s:%s:%s:%s, try "prod" next',
                    splat: [env, psm, cluster, idc, env],
                });
            }
            eps = filterEndpointsByEnv(endpoints, 'prod');
        }
        if (eps.length === 0) {
            throw new Error(`No endpoint for ${psm}:${cluster}:${idc}:${env}. ` +
                `ConsulLookupOptions: ${JSON.stringify(consulLookupOptions)}. ` +
                `Consul Agent Address: ${consul.address}. ` +
                `Endpoints before filter: ${JSON.stringify(endpoints)}. ` +
                'Please set CONSUL_HTTP_HOST or ROUTE_IP in development environment, see https://code.byted.org/nodejs/byted-service/blob/master/packages/consul/README.md#faq for details.');
        }
        const address = weightedRandom(eps);
        ctx.rip = ctx.url.host = address.addr || address.uri; // 兼容 @byted-service/consul@1.x
        resp = await ctx.fetch();
    }
    return resp;
};
exports.handle = handle;
