// @universe-design/react/src/select/common.ts
import escapeRegExp from 'lodash/escapeRegExp';
import type { Key } from 'react';
import type React from 'react';
import { useCssScope } from '../config-provider/hooks';
import type { SelectOptionGroupProps, SelectOptionProps, SelectOptionConfig } from './interface';
import type { InternalSelectOptionConfig, SelectValue } from './internal.interface';

export function useSelectListCssScope() {
  return useCssScope('select__list');
}

export function useSelectListItemCssScope() {
  return useCssScope('select__list__item');
}

export function useSelectListGroupCssScope() {
  return useCssScope('select__list__group');
}

/**
 * ======== 判断是不是我们提供的 Option 和 OptionGroup 组件 ========
 */
export const UD_SELECT_OPTION_ID = 'UD_SELECT_OPTION';
export const UD_SELECT_OPTION_GROUP_ID = 'UD_SELECT_OPTION_GROUP';
export function isSelectOption<V extends SelectValue>(node: any): node is React.ReactElement<SelectOptionProps<V>> {
  return node?.type?.[UD_SELECT_OPTION_ID];
}
export function isSelectOptionGroup<V extends SelectValue>(
  node: any,
): node is React.ReactElement<SelectOptionGroupProps<V>> {
  return node?.type?.[UD_SELECT_OPTION_GROUP_ID];
}

export const defaultFilterOption = <V extends SelectValue>(
  inputValue: string,
  option: SelectOptionConfig<V>,
): boolean => {
  const keyword = typeof option.label === 'string' ? option.label : String(option.value);
  return !inputValue || new RegExp(escapeRegExp(inputValue), 'gi').test(keyword);
};

export function getOption<V extends SelectValue, OP extends InternalSelectOptionConfig<V>>(
  value: V,
  optionMap: Map<V, OP>,
  selectedMap: Map<V, OP>,
): OP {
  // FIXME 如果没有在 option map 中找到的话做一个假的 option 返回
  return optionMap.get(value) ?? selectedMap.get(value) ?? ({ value } as OP);
}

export function getReactKey(value: SelectValue): Key {
  return typeof value === 'boolean' ? String(`${value}-bool`) : value;
}
