// @universe-design/react/src/select/index.tsx
import { COMPONENT_VERSION_ID } from '../config-provider';
import { Select as _Select } from './Select';
import { SelectAsync } from './SelectAsync';
import { SelectOption, SelectOptionGroup } from './SelectOption';

import './style';

export const Select = Object.assign(_Select, {
  Option: SelectOption,
  OptGroup: SelectOptionGroup,
  Async: SelectAsync,
  [COMPONENT_VERSION_ID]: process.env.__VERSION__,
});

export * from './selectGenerator';
export { SelectContext } from './context/SelectContext';
export type { SelectContextType } from './context/SelectContext';
export * from './interface';
export type {
  RefSelectRef,
  TagRenderer,
  SelectAction,
  MultipleSelectTagType,
  SelectValue,
  MaxTagRenderer,
  MultipleSelectMaxTagProps,
  MultipleInternalSelectProps,
  SingleInternalSelectProps,
  SelectOnSearch,
  BaseInternalSelectEventProps,
  BaseInternalSelectProps,
} from './internal.interface';
export { SelectSearchAction, SelectActionType } from './internal.interface';
export type { SelectAsyncProps } from './SelectAsync';
export { useSelectInputValue } from './context/useSelectInputValue';
