// @universe-design/react/src/select/Selector.tsx
/* eslint-disable react/destructuring-assignment */
import clsx from 'clsx';
import React, { forwardRef, useContext, useImperativeHandle, useMemo, useRef, useState } from 'react';
import stylex from '@universe-design/stylex';
import { getSizeAbbr } from '../util';
import { isDisabledOption } from './context/useSelection';
import { SelectContext } from './context/SelectContext';
import { getReactKey } from './common';
import { Overflow } from '../overflow';
import { useMultipleSelectorContext } from './context/MultipleSelectorContext';
import NativeInput from '../shared/components/Input';
import { TextOverflow } from '../shared/components/text-overflow';
import { useIsomorphicLayoutEffect as useLayoutEffect, useMouseOver } from '../hooks';
import { EmptyElement } from '../shared/components/FirstLineAlign';
import { useTextOverflowPopperSidecar } from '../config-provider/hooks/useTextOverflowPopperSidecar';
import { sxOverflowEllipsis } from '../stylex';
import { isOverflowed } from '../util/dom';
import { Text } from '../text';
import type { Context, RefObject } from 'react';
import type { SelectContextType } from './context/SelectContext';
import type {
  InternalSelectOptionConfig,
  InternalSelectProps,
  MultipleInternalSelectProps,
  SelectValue,
} from './internal.interface';

const Input = forwardRef<HTMLElement>(function SelectorInput(props, ref) {
  const {
    id,
    inputRef,
    selectorCssScope: cssScope,
    inputValue,
    disabled,
    onInputChange,
    onInputCompositionStart,
    onInputCompositionEnd,
    onInputPaste,
    showSearch,
    tabIndex,
    inputting,
  } = useContext(SelectContext);

  useImperativeHandle<HTMLElement | null, HTMLElement | null>(ref, () => inputRef.current);

  return (
    <NativeInput
      id={id}
      tabIndex={tabIndex}
      role="combobox"
      autoComplete="off"
      disabled={disabled}
      type="search"
      readOnly={!showSearch}
      ref={inputRef as RefObject<HTMLInputElement>}
      className={clsx(cssScope('__search__input'), inputting && cssScope('__search__input-inputting'))}
      value={inputValue}
      onChange={onInputChange}
      onPaste={onInputPaste}
      onCompositionStart={onInputCompositionStart}
      onCompositionEnd={onInputCompositionEnd}
    />
  );
});

const AutoSizeInput = () => {
  const { selectorCssScope: cssScope, inputValue } = useContext(SelectContext);

  const measureRef = useRef<HTMLSpanElement>(null);
  const [inputWidth, setInputWidth] = useState(0);

  useLayoutEffect(() => {
    if (measureRef.current) {
      setInputWidth(measureRef.current.offsetWidth);
    }
  }, [inputValue]);

  return (
    <span
      className={cssScope('__search__input_container')}
      style={{
        // TODO 不优雅
        width: inputWidth + 2,
      }}
    >
      <Input />
      <span ref={measureRef} className={cssScope('__search__measure')} aria-hidden>
        {inputValue}&nbsp;
      </span>
    </span>
  );
};

export function MultipleSelector<V extends SelectValue, OP extends InternalSelectOptionConfig<V>>({
  maxTagCount,
  maxTagRenderer,
}: Required<Pick<MultipleInternalSelectProps<V, OP>, 'maxTagRenderer'>> &
  Pick<MultipleInternalSelectProps<V, OP>, 'maxTagCount'>) {
  const {
    valueSet,
    getOption,
    deselect,
    disabled: globalDisabled,
    selectorCssScope: cssScope,
    selectorLabelRenderer,
    tagRenderer,
  } = React.useContext(SelectContext as unknown as Context<SelectContextType<V, OP>>);

  const autoSizeInputNode = useMemo(() => <AutoSizeInput />, []);

  return (
    <Overflow
      className={clsx(cssScope('__search'))}
      maxCount={maxTagCount}
      suffixClassName={cssScope('__search__suffix')}
      suffixWrap={false}
      keys={[...valueSet].map(getReactKey)}
      itemRenderer={opt => {
        const _opt = opt as V;
        const option = getOption(_opt);
        const disabled = isDisabledOption(option, globalDisabled);
        return tagRenderer({
          title: option.htmlTitle,
          value: option.value,
          disabled,
          closable: !disabled,
          onClose: () => {
            deselect(_opt);
          },
          children: selectorLabelRenderer(_opt),
        });
      }}
      restRenderer={omittedKeys =>
        maxTagRenderer(
          {
            disabled: globalDisabled,
          },
          omittedKeys.map(opt => {
            const _opt = opt as V;
            const option = getOption(_opt);
            const disabled = isDisabledOption(option, globalDisabled);
            return {
              value: option.value,
              disabled,
              closable: !disabled,
              onClose: () => {
                deselect(_opt);
              },
              children: selectorLabelRenderer(_opt),
            };
          }),
        )
      }
      suffixRenderer={() => autoSizeInputNode}
    />
  );
}

export function SingleSelector() {
  const { selectorCssScope: cssScope } = React.useContext(SelectContext);

  return (
    <div className={cssScope('__search')}>
      <Input />
    </div>
  );
}

type SelectorProps<V extends SelectValue, OP extends InternalSelectOptionConfig<V>> = Required<
  Pick<
    InternalSelectProps<V, OP>,
    'placeholder' | 'size' | 'bordered' | 'error' | 'allowClear' | 'showArrow' | 'clearIcon' | 'suffixIcon'
  > &
    Pick<MultipleInternalSelectProps<V, OP>, 'maxTagRenderer'>
> &
  Pick<MultipleInternalSelectProps<V, OP>, 'maxTagCount'>;

function SelectorForwardRefRenderFunction<V extends SelectValue, OP extends InternalSelectOptionConfig<V>>(
  {
    size,
    placeholder,
    bordered,
    error,
    maxTagCount,
    allowClear,
    showArrow,
    maxTagRenderer,
    clearIcon,
    suffixIcon,
  }: SelectorProps<V, OP>,
  ref?: React.Ref<HTMLDivElement>,
) {
  const {
    multiple,
    focusState,
    overlayVisible,
    clear,
    valueSet,
    inputValue,
    disabled,
    selectorCssScope: cssScope,
    selectorLabelRenderer,
    setOverlayVisible,
    inputRef,
    inputting,
    showSearch,
  } = React.useContext(SelectContext as unknown as Context<SelectContextType<V, OP>>);
  const { maxHeight: multipleSelectorMaxHeight } = useMultipleSelectorContext();

  const { isMouseOver, handleMouseEnter, handleMouseLeave } = useMouseOver();
  const TextOverflowPopper = useTextOverflowPopperSidecar();
  const placeholderRef = React.useRef<HTMLDivElement>(null);
  const [placeholderTooltipVisible, setPlaceholderTooltipVisible] = React.useState(false);

  let selectorNode;
  if (multiple) {
    selectorNode = <MultipleSelector maxTagCount={maxTagCount} maxTagRenderer={maxTagRenderer} />;
  } else {
    selectorNode = <SingleSelector />;
  }

  const isSelectValueEmpty = !valueSet.size;
  const textVariant = (
    {
      small: 'body2',
      medium: 'body0',
      large: 'title5',
    } as const
  )[size];
  // 是否应该显示 placeholder
  const isPlaceholderVisible = isSelectValueEmpty && Boolean(placeholder) && !inputValue;
  /**
   * 这儿加个 key
   * 在切换 placeholder 和 selectItem 时, react 有可能会复用节点
   * 有可能切换到 selectItem 的一瞬间, 由于还是复用的 placeholder 的节点
   * 所以那一瞬间继承了 placeholder 的 class, 导致文字颜色有一瞬间是 placeholder 的颜色
   * 此时如果这个节点上有个 transition 的过渡, 就会导致该节点的颜色闪一下
   */
  let placeholderNode: React.ReactNode = (
    <Text
      key="placeholder"
      as="div"
      variant={textVariant}
      className={clsx(cssScope('__placeholder'), stylex(sxOverflowEllipsis))}
      ref={placeholderRef}
    >
      {placeholder}
    </Text>
  );

  if (inputValue) {
    if (multiple && !isSelectValueEmpty) {
      placeholderNode = null;
    } else {
      placeholderNode = <span className={cssScope('__placeholder')} />;
    }
  } else if (!isSelectValueEmpty) {
    if (multiple) {
      placeholderNode = null;
    } else {
      // first item
      const selectItem = valueSet.keys().next().value!;
      placeholderNode = (
        <TextOverflow
          key="selectItem"
          as="div"
          className={clsx(cssScope('__selectItem'), inputting && cssScope('__selectItem-inputting'))}
          lineClamp={1}
          variant={textVariant}
          disabled={disabled}
        >
          {selectorLabelRenderer(selectItem)}
        </TextOverflow>
      );
    }
  }
  const showClear = !isSelectValueEmpty && !disabled && allowClear;

  /**
   * 兼容其他 DS
   */
  function getBorderStyle(): Exclude<typeof bordered, boolean> {
    if (bordered === true) {
      return 'normal';
    } else if (bordered === false) {
      return 'noborder';
    }
    return bordered;
  }

  const isPlaceholderOverflowed = () => (placeholderRef.current ? isOverflowed(placeholderRef.current) : false);

  return (
    <div
      ref={ref}
      onMouseEnter={handleMouseEnter}
      onMouseLeave={handleMouseLeave}
      className={clsx(
        cssScope(),
        cssScope(`-${getSizeAbbr(size)}`),
        focusState && cssScope('-focus'),
        overlayVisible && cssScope('-open'),
        !showSearch && cssScope('-readOnly'),
        cssScope(`-border-${getBorderStyle()}`),
        !isSelectValueEmpty && cssScope('-not-empty'),
        showClear && cssScope('-allowClear'),
        // isMouseOver && showClear && cssScope('-showClear'),
        multiple && cssScope('-multiple'),
        disabled && cssScope('-disabled'),
        error && cssScope('-error'),
        !showArrow && cssScope('-hide-arrow'),
      )}
      onClick={({ target }) => {
        // 这里要用 onClick 不能用 mousedown 和 mouseup，因为 multiple 模式下 tag 的 onClick 做了 stopPropagation
        if (target !== inputRef.current) {
          inputRef.current?.focus();
        }
        setOverlayVisible();
      }}
      style={{
        maxHeight: multiple ? multipleSelectorMaxHeight : undefined,
      }}
    >
      <EmptyElement
        sx={[
          {
            small: stylex.create({ lineHeight: '24px' }),
            medium: stylex.create({ lineHeight: '28px' }),
            large: stylex.create({ lineHeight: '36px' }),
          }[size],
        ]}
      />
      <TextOverflowPopper
        title={<>{placeholder}</>}
        checkOverflow={isPlaceholderOverflowed}
        visible={isPlaceholderVisible && placeholderTooltipVisible}
        onVisibleChange={visible => {
          setPlaceholderTooltipVisible(visible);
        }}
      >
        <div
          className={cssScope('__content')}
          onMouseLeave={() => {
            setPlaceholderTooltipVisible(false);
          }}
        >
          {/* showSearch 的时候 input 会盖住 placeholder 所以 tooltip 是不会弹出的，这个符合预期 */}
          {placeholderNode}
          {/* 目前不能直接在 showSearch false 的时候删除 selectorNode，因为我们会利用 input 的 focus 状态来做键盘事件 */}
          {selectorNode}
        </div>
      </TextOverflowPopper>
      {showClear ? (
        <div
          className={clsx(
            cssScope('__clear'),
            stylex(isMouseOver ? stylex.create({ opacity: 1 }) : stylex.create({ opacity: 0 })),
          )}
          onMouseDown={e => {
            e.stopPropagation();
            clear();
          }}
          onClick={e => {
            // 防止点击冒泡触发 overlay 打开
            e.stopPropagation();
          }}
        >
          {clearIcon}
        </div>
      ) : null}
      {showArrow ? <div className={clsx(cssScope('__arrow'))}>{suffixIcon}</div> : null}
    </div>
  );
}

interface SelectorForwardRefRenderFunctionType {
  <V extends SelectValue, OP extends InternalSelectOptionConfig<V>>(
    props: SelectorProps<V, OP>,
  ): React.ReactElement | null;
}

// TODO typing is not safe!
export const Selector = React.forwardRef(SelectorForwardRefRenderFunction) as SelectorForwardRefRenderFunctionType;
