// @universe-design/react/src/select/SelectAsync.tsx
import type { PropsWithChildren, PropsWithoutRef, ReactElement, Reducer, Ref, RefAttributes } from 'react';
import React, { forwardRef, useCallback, useEffect, useReducer } from 'react';
import type { Observable } from 'rxjs';
import { BehaviorSubject, from, isObservable, of, Subject, throwError, timer } from 'rxjs';
import { switchMap, debounce, withLatestFrom, tap, map, catchError } from 'rxjs/operators';
import { Select } from './Select';
import { useConstant } from '@universe-design/react-utilities';
import { useMountedState, useEffectOnce } from '../hooks';
import { useLoading } from '../loading/useLoading';
import { devWarning, isFunction, isPromise, warning } from '../util';
import type { SelectProps, SelectOptionConfig, SelectOptionGroupConfig, SelectRef } from './interface';
import type { SelectValue } from './internal.interface';

type SelectOptionsType<V extends SelectValue> = (SelectOptionConfig<V> | SelectOptionGroupConfig<V>)[];

export interface SelectAsyncProps<V extends SelectValue> {
  /**
   * 在用户搜索前默认显示的选项集合。当为 `true` 时，会自动调用 `loadOptions('', [])` 获取数据，当为 `false` 时什么都不做，当为数组时，直接用选项数组作为默认值。
   */
  defaultOptions?: SelectOptionsType<V> | boolean;
  /**
   * 异步加载选项函数，可以返回 `Promise`，`Observable` 或直接返回选项数组
   */
  loadOptions?: (
    inputValue: string,
    prevOptions: SelectOptionsType<V>,
  ) => Promise<SelectOptionsType<V>> | Observable<SelectOptionsType<V>> | SelectOptionsType<V>;
  /**
   * 远程搜索的去抖动时间
   * @defaultValue 800
   */
  debounceTime?: number;
  /**
   * 加载状态出现的延迟时间，用于避免闪动
   * @defaultValue 200
   */
  loadingDelay?: number;
  /**
   * 开始加载选项前的事件回调，区别于 `onSearch`，此回调在初次渲染加载前也会调用
   */
  onBeforeLoadOptions?: () => void;
  /**
   * 加载选项成功的事件回调
   */
  onLoadOptionsSuccess?: (data: SelectOptionsType<V>) => void;
  /**
   * 加载选项失败的事件回调
   */
  onLoadOptionsError?: (err: any) => void;
}

enum Actions {
  StartFetching,
  GotOptionsSuccessfully,
  GotOptionsFailed,
}

export function useSelectAsync<V extends SelectValue = string>(props: SelectProps<V> & SelectAsyncProps<V>) {
  const {
    defaultOptions = [],
    loadOptions,
    onSearch,
    debounceTime = 800,
    loadingDelay = 200,
    onBeforeLoadOptions,
    onLoadOptionsError,
    onLoadOptionsSuccess,
    // omit
    options: optionsProp,
    filterOption: filterOptionProp,
    loading: loadingProp,
    ...restProps
  } = props;

  warning(optionsProp === undefined, devWarning('useSelectAsync', 'options will be ignored in async select'));

  warning(loadingProp === undefined, devWarning('useSelectAsync', 'loading will be ignored in async select'));

  warning(filterOptionProp === undefined, devWarning('useSelectAsync', 'filterOption will be ignored in async select'));

  const initialState = useConstant(() => ({
    loading: false,
    options: Array.isArray(defaultOptions) ? defaultOptions : [],
    searchValue: restProps.searchValue ?? '',
  }));

  const [state, dispatch] = useReducer<
    Reducer<
      {
        loading: boolean;
        options: SelectOptionsType<V>;
        error?: any;
      },
      | { type: Actions.StartFetching }
      | { type: Actions.GotOptionsSuccessfully; payload: SelectOptionsType<V> }
      | { type: Actions.GotOptionsFailed; payload: any }
    >
  >((_state, action) => {
    switch (action.type) {
      case Actions.StartFetching:
        isFunction(onBeforeLoadOptions) && onBeforeLoadOptions();
        return {
          ..._state,
          loading: true,
        };
      case Actions.GotOptionsSuccessfully:
        isFunction(onLoadOptionsSuccess) && onLoadOptionsSuccess(action.payload);
        return {
          ..._state,
          options: action.payload,
          loading: false,
        };
      case Actions.GotOptionsFailed:
        isFunction(onLoadOptionsError) && onLoadOptionsError(action.payload);
        return {
          ..._state,
          loading: false,
          error: action.payload,
        };
      default:
        throw new Error('Action type not found');
    }
  }, initialState);
  const [delayedLoading] = useLoading({
    loading: state.loading,
    delay: loadingDelay,
  });
  const isMounted = useMountedState();
  // const isFirstMount = useFirstMountState();

  // Observable
  const event$ = useConstant(
    () =>
      new Subject<{
        searchValue: string;
        showLoading: boolean;
        delay: number;
      }>(),
  );
  const state$ = useConstant(
    () =>
      new BehaviorSubject<{
        options: SelectOptionsType<V>;
      }>(initialState),
  );
  useEffect(() => {
    const value$ = event$.pipe(
      debounce(ev => (ev.delay ? timer(ev.delay) : of(ev))),
      withLatestFrom(state$),
      tap(([ev]) => {
        if (isMounted() && ev.showLoading) {
          dispatch({ type: Actions.StartFetching });
        }
      }),
      switchMap(([ev, _state]) => {
        if (!isFunction(loadOptions)) {
          return of(_state);
        }
        try {
          const ret = loadOptions(ev.searchValue, _state.options);
          const mapOptions = map((_options: SelectOptionsType<V>) => ({
            ..._state,
            options: _options,
          }));
          // eslint-disable-next-line no-nested-ternary
          return (isObservable(ret) ? ret : isPromise(ret) ? from(ret) : of(ret)).pipe(
            mapOptions,
            catchError(err => {
              console.error('Unexpected error occured when loading options: ', err, 'Please handle it manually');
              if (isMounted()) {
                dispatch({ type: Actions.GotOptionsFailed, payload: err });
              }
              return of(_state);
            }),
          );
        } catch (err) {
          return throwError(err);
        }
      }),
    );
    const subscription = value$.subscribe(
      value => {
        if (isMounted()) {
          state$.next(value);
          dispatch({
            type: Actions.GotOptionsSuccessfully,
            payload: value.options,
          });
        }
      },
      err => {
        console.log('value$ complete with error: ', err);
      },
    );
    return () => {
      subscription.unsubscribe();
    };
  }, [event$, isMounted, loadOptions, state$]); // immutable forever

  useEffectOnce(() => {
    if (defaultOptions === false || Array.isArray(defaultOptions)) {
      return;
    }
    if (isFunction(loadOptions)) {
      event$.next({
        searchValue: initialState.searchValue,
        delay: 0,
        showLoading: false,
      });
    }

    return () => {
      state$.complete();
      event$.complete();
    };
  });

  const handleSearch: NonNullable<typeof onSearch> = useCallback(
    (v, meta) => {
      event$.next({
        searchValue: v,
        delay: debounceTime,
        showLoading: true,
      });
      isFunction(onSearch) && onSearch(v, meta);
    },
    [debounceTime, event$, onSearch],
  );

  const additionalProps: SelectProps<V> = {
    loading: delayedLoading,
    options: state.options,
    onSearch: handleSearch,
    filterOption: false,
  };

  return {
    ...additionalProps,
    ...restProps,
  };
}

export const SelectAsync = forwardRef(
  <V extends SelectValue>(props: PropsWithChildren<SelectProps<V> & SelectAsyncProps<V>>, ref: Ref<SelectRef<V>>) => {
    const asyncProps = useSelectAsync(props);
    // FIXME error TS2769: No overload matches this call.
    return <Select ref={ref} {...(asyncProps as any)} />;
  },
) as {
  <V extends SelectValue>(
    props: PropsWithoutRef<PropsWithChildren<SelectProps<V> & SelectAsyncProps<V>>> & RefAttributes<SelectRef<V>>,
  ): ReactElement;
};
