// @universe-design/react/src/select/Select.tsx
import ListCheckBoldOutlined from '@universe-design/icons-react/ListCheckBoldOutlined';
import type { PropsWithChildren, PropsWithoutRef, ReactElement, Ref, RefAttributes } from 'react';
import React from 'react';
import { useForceUpdate } from '@universe-design/react-utilities';
import { isNonNullable, toFlatArray } from '../util';
import { isSelectOption, isSelectOptionGroup } from './common';
import type {
  SelectProps,
  SelectOptionConfig,
  SelectOptionGroupConfig,
  SingleSelectProps,
  MultipleSelectProps,
  RefOptionListProps,
  SelectRef,
} from './interface';
import type { InternalSelectOptionConfig, SelectValue } from './internal.interface';
import { selectGenerator } from './selectGenerator';
import { SelectOverlay } from './SelectOverlay';

const RefSelect = selectGenerator<SelectValue, SelectOptionConfig<SelectValue>>({
  selectorCssScope: 'select',
  dropdownCssScope: 'select',
  defaultOptionLabelProp: option => option.labelRenderer?.('') ?? option.label ?? option.value,
});

function selectChildren2Options<V extends SelectValue>(
  children: React.ReactNode,
): NonNullable<SelectProps<V>['options']> {
  return toFlatArray(children)
    .map((child, index) => {
      if (isSelectOption<V>(child)) {
        const {
          key,
          props: { children: _children, ...restProps },
        } = child;
        return {
          ...restProps,
          label: _children,
          key: key ?? restProps.value ?? `select-option-key-${index}`,
        } as SelectOptionConfig<V>;
      } else if (isSelectOptionGroup(child)) {
        const {
          key,
          props: { children: _children, ...restProps },
        } = child;
        return {
          ...restProps,
          key: key || `select-optgroup-key-${index}`,
          options: selectChildren2Options(_children),
        } as SelectOptionGroupConfig<V>;
      }
      return null;
    })
    .filter(isNonNullable);
}

export const Select = React.forwardRef(
  <V extends SelectValue = string>(props: PropsWithChildren<SelectProps<V>>, ref: Ref<SelectRef<V>>) => {
    const {
      options = selectChildren2Options<V>(props.children),
      filterOption = true,
      defaultActiveFirstOption = true,
      menuItemSelectedIcon = <ListCheckBoldOutlined />,
      listHeight = 248,
      onPopupScroll,
      virtual = true,
      itemHeight = 32,
      filterSort,
      debounceTime,
      showSelectAll,
      onActiveOptionChange,
      ...baseSelectProps
    } = props;

    const internalOptions = React.useMemo(
      () =>
        options.reduce((pre, cur) => {
          if ('options' in cur) {
            for (const opt of cur.options) {
              pre.push(Object.assign({}, opt, { htmlTitle: opt.title }));
            }
          } else {
            pre.push(Object.assign({}, cur, { htmlTitle: cur.title }));
          }
          return pre;
        }, [] as InternalSelectOptionConfig<V>[]),
      [options],
    );

    const forceUpdate = useForceUpdate();
    // We need force update here since popup dom is render async
    function onPopupMouseEnter() {
      forceUpdate();
    }

    const refSelectRef = React.useRef<SelectRef<V>>(null);
    const listRef = React.useRef<RefOptionListProps<V>>(null);
    React.useImperativeHandle(
      ref,
      () => ({
        get input() {
          return refSelectRef.current?.input;
        },
        focus() {
          return refSelectRef.current?.focus();
        },
        blur() {
          return refSelectRef.current?.blur();
        },
        // deprecated: use scrollInfoView instead
        scrollTo(...args) {
          return listRef.current?.scrollIntoView(...args);
        },
        scrollIntoView(...args) {
          return listRef.current?.scrollIntoView(...args);
        },
        setActiveOption(value) {
          return listRef.current?.setActiveOption(value);
        },
      }),
      [],
    );

    return (
      <RefSelect<V, SelectOptionConfig<V>>
        {...baseSelectProps}
        ref={refSelectRef}
        options={internalOptions}
        overlayNode={
          <SelectOverlay<V>
            ref={listRef}
            filterSort={filterSort}
            defaultActiveFirstOption={defaultActiveFirstOption}
            options={options}
            filterOption={filterOption}
            menuItemSelectedIcon={menuItemSelectedIcon}
            listHeight={listHeight}
            onPopupScroll={onPopupScroll}
            virtual={virtual}
            itemHeight={itemHeight}
            onMouseEnter={onPopupMouseEnter}
            onActiveOptionChange={onActiveOptionChange}
            debounceTime={debounceTime}
            showSelectAll={showSelectAll}
          />
        }
      />
    );
  },
) as {
  <V extends SelectValue = string>(
    props: PropsWithoutRef<PropsWithChildren<MultipleSelectProps<V>>> & RefAttributes<SelectRef<V>>,
  ): ReactElement | null;
  <V extends SelectValue = string>(
    // eslint-disable-next-line @typescript-eslint/unified-signatures
    props: PropsWithoutRef<PropsWithChildren<SingleSelectProps<V>>> & RefAttributes<SelectRef<V>>,
  ): ReactElement | null;
};
