// @universe-design/react/src/select/SelectOverlay.tsx
import noop from 'lodash/noop';
import * as React from 'react';
import type { ListRef } from '@universe-design/rc-virtual-list';
import List from '@universe-design/rc-virtual-list';
import { isNonNullable } from '../util';
import { KeyboardKey } from '@universe-design/react-utilities';
import { defaultFilterOption, getReactKey, useSelectListCssScope } from './common';
import type {
  RefOptionListProps,
  SelectFilterOptionFunc,
  SelectOptionConfig,
  SelectOptionGroupConfig,
  SelectOverlayProps,
  SelectProps,
} from './interface';
import { SelectContext } from './context/SelectContext';
import { InternalSelectOption, InternalSelectOptionGroup } from './SelectOption';
import type { Required } from 'utility-types';
import { useEffectOnce, useDebounce, useLatest, useEventCallback } from '../hooks';
import type { SelectOnActiveOptionChange, SelectValue } from './internal.interface';
import { SelectActiveOptionChangeAction } from './internal.interface';
import composeRefs from '../util/composeRefs';
import { ListDivider, ListItemLike } from '../shared/components/list';
import { Checkbox } from '../checkbox';
import useSelectAll from './context/useSelectAll';
import { useLocale } from '../config-provider/hooks';
import { SelectOverlayContainer } from './SelectOverlayContainer';

// 这里和 option shape 做一下区分
interface SimpleFlatternOption<V extends SelectValue> {
  group: false;
  option: SelectOptionConfig<V>;
}
interface GroupFlatternOption<V extends SelectValue> {
  group: true;
  option: SelectOptionGroupConfig<V>;
}
export type FlatternOption<V extends SelectValue> = GroupFlatternOption<V> | SimpleFlatternOption<V>;

function isGroupFlatternOption<V extends SelectValue>(opt: FlatternOption<V>): opt is GroupFlatternOption<V> {
  return opt.group;
}

function isSimpleFlatternOption<V extends SelectValue>(opt: FlatternOption<V>): opt is SimpleFlatternOption<V> {
  return !opt.group;
}

function getFirstOption<V extends SelectValue>(
  options: FlatternOption<V>[],
  predicate: (value: SimpleFlatternOption<V>) => boolean = () => true,
) {
  for (const opt of options) {
    if (isSimpleFlatternOption(opt) && !opt.option.disabled && predicate(opt)) {
      return opt.option;
    }
  }
}

function getLastOption<V extends SelectValue>(options: FlatternOption<V>[]) {
  const opt = options
    .filter(isSimpleFlatternOption)
    .filter(o => !o.option.disabled)
    .pop();
  if (!opt) {
    return undefined;
  }
  return opt.option;
}

export function getNextActiveOption<V extends SelectValue>(
  activeOption: SelectOptionConfig<V> | undefined,
  options: FlatternOption<V>[],
  forward: boolean,
): SelectOptionConfig<V> | undefined {
  const curActiveKeyIndex = options.findIndex(
    opt => isSimpleFlatternOption(opt) && !opt.option.disabled && opt.option.value === activeOption?.value,
  );
  if (curActiveKeyIndex === -1) {
    if (forward) {
      return getFirstOption(options);
    } else {
      return getLastOption(options);
    }
  }

  for (
    let i = curActiveKeyIndex, count = 0;
    count < options.length;
    count += 1,
      // eslint-disable-next-line no-nested-ternary
      i = forward // 正向遍历或反向遍历整个数组
        ? i + 1 > options.length - 1
          ? 0
          : i + 1
        : i - 1 < 0
        ? options.length - 1
        : i - 1
  ) {
    if (i === curActiveKeyIndex) {
      continue;
    }
    const opt = options[i];
    if (isSimpleFlatternOption(opt) && !opt.option.disabled) {
      return opt.option;
    }
  }
}

export function useSelectOverlay<V extends SelectValue>(
  props: Required<
    Pick<
      SelectProps<V>,
      | 'options'
      | 'filterOption'
      | 'defaultActiveFirstOption'
      | 'filterSort'
      | 'debounceTime'
      | 'showSelectAll'
      | 'onActiveOptionChange'
    >,
    'options' | 'filterOption' | 'defaultActiveFirstOption' | 'debounceTime'
  > & {
    ref:
      | ((instance: RefOptionListProps<V> | null) => void)
      | React.MutableRefObject<RefOptionListProps<V> | null>
      | null;
  },
) {
  const {
    inputValue: ctxInputValue,
    overlayVisible,
    valueSet,
    toggle,
    listRef,
    compositionStatusRef,
    set,
    disabled,
  } = React.useContext(SelectContext);
  const {
    defaultActiveFirstOption = true,
    options,
    filterOption,
    filterSort,
    debounceTime,
    onActiveOptionChange,
  } = props;

  const [inputValue] = useDebounce(ctxInputValue, debounceTime);

  // =========================== List ===========================
  const flattenOptions = React.useMemo(() => {
    const _filterOption: SelectFilterOptionFunc<V> = (_inputValue, _optProps) => {
      if (filterOption === true) {
        return defaultFilterOption(_inputValue, _optProps);
      } else if (filterOption === false) {
        return true;
      }
      return filterOption(_inputValue, _optProps);
    };

    const shouldSort = filterSort && inputValue;

    /**
     * 只筛选，保留树形结构
     */
    const filteredOptions = options
      .map(optProps => {
        if (!('options' in optProps) && !_filterOption(inputValue, optProps)) {
          return null;
        }
        if ('options' in optProps) {
          const _filteredOptions = optProps.options.filter(_optProps => _filterOption(inputValue, _optProps));
          if (_filteredOptions.length) {
            return {
              ...optProps,
              options: shouldSort ? _filteredOptions.sort(filterSort) : _filteredOptions,
            };
          }
          return null;
        }
        return optProps;
      })
      .filter(isNonNullable);

    if (shouldSort) {
      filteredOptions.sort(filterSort);
    }

    /**
     * 打平树形结构
     */
    const res: FlatternOption<V>[] = [];
    for (const opt of filteredOptions) {
      if ('options' in opt) {
        res.push({ group: true, option: opt });
        for (const _opt of opt.options) {
          res.push({ group: false, option: _opt });
        }
      } else {
        res.push({ group: false, option: opt });
      }
    }

    return res;
  }, [filterOption, filterSort, inputValue, options]);
  const latestFlattenOptionsRef = useLatest(flattenOptions);

  const { onToggleSelectAll, hasSelectedAll, isSelectAllDisabled, selectAllCount } = useSelectAll({
    set,
    flattenOptions,
    valueSet,
  });
  // =========================== List ===========================
  const rcVirtualListRef = React.useRef<ListRef>(null);

  /**
   * 滚动到特定选项，支持按照三种方式滚动：
   *  1. index: 根据选项所在 index
   *  2. option value
   *  3. groupKey
   */
  const scrollIntoView: RefOptionListProps<V>['scrollIntoView'] = (reference, scrollConfig) => {
    const list = rcVirtualListRef.current;
    if (!list) {
      return;
    }

    let index = -1;
    if (typeof reference === 'number') {
      index = reference;
    } else if (reference.type === 'option') {
      index = flattenOptions.findIndex(
        flattenOption => !flattenOption.group && flattenOption.option.value === reference.value,
      );
    } else if (reference.type === 'group') {
      index = flattenOptions.findIndex(
        flattenOption => flattenOption.group && flattenOption.option.key === reference.key,
      );
    }

    if (index !== -1) {
      list.scrollTo({
        index,
        ...scrollConfig,
      });
    }
  };

  // ========================== Active ==========================
  /**
   * 逻辑
   * - 优先级
   *   - active 上次关闭时的项
   *   - active 第一个选中项
   *   - active 第一个选项
   *   - defaultActiveFirstOption false
   * - 场景
   *   - 首次打开
   *   - 再次打开
   *   - 选项列表变化，比如搜索，或异步选取等
   */
  const firstOption = React.useMemo(() => getFirstOption(flattenOptions), [flattenOptions]);
  const latestFirstOptionRef = useLatest(firstOption);

  const firstSelectedOption = React.useMemo(
    () => getFirstOption(flattenOptions, opt => valueSet.has(opt.option.value)),
    [flattenOptions, valueSet],
  );

  const defaultOption = React.useMemo(() => {
    if (!defaultActiveFirstOption) {
      return undefined;
    }
    return firstSelectedOption ?? firstOption;
  }, [defaultActiveFirstOption, firstOption, firstSelectedOption]);

  const [activeOption, _setActiveOption] = React.useState<SelectOptionConfig<V> | undefined>(defaultOption);

  const setActiveOption = React.useCallback(
    (meta: Parameters<SelectOnActiveOptionChange<V>>[1]) => {
      onActiveOptionChange && onActiveOptionChange(meta.option?.value, meta);
      _setActiveOption(meta.option);
    },
    [onActiveOptionChange],
  );
  const latestSetActiveOption = useEventCallback(setActiveOption);

  // 初次进入滑到对应的元素
  useEffectOnce(() => {
    scrollIntoView(flattenOptions.findIndex(o => isSimpleFlatternOption(o) && o.option.value === activeOption?.value));
  });

  // ========================= Keyboard =========================
  React.useImperativeHandle(composeRefs(listRef, props.ref), () => ({
    onKeyDown: e => {
      if (e.isDefaultPrevented()) {
        return;
      }

      let up: boolean;

      function setActiveAndScrollTo() {
        const nextOption = getNextActiveOption(activeOption, flattenOptions, !up);
        if (nextOption) {
          // 跟随 active 元素滚动
          setActiveOption({
            option: nextOption,
            action: up ? SelectActiveOptionChangeAction.ArrowUp : SelectActiveOptionChangeAction.ArrowDown,
          });
          scrollIntoView(
            flattenOptions.findIndex(o => isSimpleFlatternOption(o) && o.option.value === nextOption.value),
          );
        }
      }

      switch (e.key) {
        case KeyboardKey.Down: // IE/Edge specific value
        case KeyboardKey.ArrowDown:
          up = false;
          setActiveAndScrollTo();
          break;
        case KeyboardKey.Up: // IE/Edge specific value
        case KeyboardKey.ArrowUp:
          up = true;
          setActiveAndScrollTo();
          break;

        // >>> Select
        case KeyboardKey.Enter: // overlay 操作选中值或 toggle overlay
          if (activeOption && overlayVisible && !compositionStatusRef.current) {
            e.preventDefault();
            toggle(activeOption.value);
          }
          break;
        default:
          return;
      }
    },
    onKeyUp: noop,
    scrollTo: scrollIntoView,
    scrollIntoView,
    setActiveOption: value => {
      setActiveOption({
        option: flattenOptions.filter(isSimpleFlatternOption).find(o => o.option.value === value)?.option,
        action: SelectActiveOptionChangeAction.Manual,
      });
    },
  }));

  const isFirstMount = React.useRef(true);
  React.useEffect(() => {
    if (isFirstMount.current) {
      isFirstMount.current = false;
      return;
    }
    // 这里限制只有 inputValue 变了才触发重新设置 active option。因为 option list 还可能因为操作有新增和删除，这时候变化 active，可能有超出预期的情况。规范上最明确的预期就是搜索后，尽量保证有 active 选项，排除其他更高优逻辑，就是首个选项 active 兜底，所以限定在搜索时触发就行了 -- @qiuchen.1001
    inputValue;

    latestSetActiveOption({
      option: getFirstOption(
        latestFlattenOptionsRef.current,
        opt => opt.option.value === latestFirstOptionRef.current?.value,
      ),
      action: SelectActiveOptionChangeAction.AfterSearch,
    });
  }, [latestFirstOptionRef, latestFlattenOptionsRef, inputValue, latestSetActiveOption]);

  return {
    activeOption,
    setActiveOption,
    rcVirtualListRef,
    flattenOptions,
    onToggleSelectAll,
    hasSelectedAll,
    isSelectAllDisabled,
    selectAllCount,
    disabled,
  };
}

export const SelectOverlay = React.forwardRef(function SelectOverlay<V extends SelectValue>(
  props: SelectOverlayProps<V>,
  ref: React.Ref<RefOptionListProps<V>>,
) {
  const {
    options,
    filterOption,
    defaultActiveFirstOption,
    menuItemSelectedIcon,
    listHeight,
    onPopupScroll,
    itemHeight,
    virtual,
    onMouseEnter,
    filterSort,
    debounceTime = 0,
    showSelectAll,
    onActiveOptionChange,
  } = props;
  const cssScope = useSelectListCssScope();

  const {
    activeOption,
    setActiveOption,
    rcVirtualListRef,
    flattenOptions,
    hasSelectedAll,
    onToggleSelectAll,
    isSelectAllDisabled,
    selectAllCount,
    disabled,
  } = useSelectOverlay({
    ref,
    options,
    filterOption,
    defaultActiveFirstOption,
    filterSort,
    debounceTime,
    onActiveOptionChange,
  });

  const local = useLocale();
  // ========================== Render ==========================
  const menuNode = (
    <div className={cssScope('__menu')}>
      <List<FlatternOption<V>>
        itemKey={option =>
          isGroupFlatternOption(option) ? `group-${option.option.key}` : getReactKey(option.option.value)
        }
        ref={rcVirtualListRef}
        data={flattenOptions}
        height={listHeight - 3 * 2 - 1 * 2}
        itemHeight={itemHeight}
        fullHeight={false}
        onScroll={onPopupScroll}
        virtual={virtual}
        onMouseEnter={onMouseEnter}
      >
        {(option, itemIndex) => {
          // Group
          if (isGroupFlatternOption(option)) {
            return (
              <InternalSelectOptionGroup
                {...{
                  ...option.option,
                  isFirst: itemIndex === 0,
                }}
              />
            );
          }

          // Option
          const { option: origOpt } = option;
          return (
            // 这里加一个空 div 是为了把内部带 margin 的元素包住，让 virtual list 能正常计算，tree，select，cascader
            <div>
              <InternalSelectOption
                active={origOpt.value === activeOption?.value}
                key={getReactKey(origOpt.value)}
                menuItemSelectedIcon={menuItemSelectedIcon}
                {...origOpt}
                onMouseMove={() =>
                  setActiveOption({
                    option: origOpt,
                    action: SelectActiveOptionChangeAction.MouseMove,
                  })
                }
              />
            </div>
          );
        }}
      </List>
      {showSelectAll && (
        <>
          <ListDivider fluid />
          <ListItemLike>
            <Checkbox
              disabled={disabled || isSelectAllDisabled}
              checked={hasSelectedAll}
              onChange={onToggleSelectAll}
              data-testid="select-all"
            >
              {local.global.selectAll} ({selectAllCount})
            </Checkbox>
          </ListItemLike>
        </>
      )}
    </div>
  );

  return (
    <SelectOverlayContainer
      {...{
        optionsLength: options.length,
        filteredOptionsLength: flattenOptions.length,
        menu: menuNode,
      }}
    />
  );
}) as <V extends SelectValue>(
  props: React.PropsWithoutRef<SelectOverlayProps<V>> & React.RefAttributes<RefOptionListProps<V>>,
) => React.ReactElement | null;
