// @universe-design/react/src/select/SelectOption.tsx
import clsx from 'clsx';
import type { PropsWithChildren, Ref } from 'react';
import React, { forwardRef } from 'react';
import stylex from '@universe-design/stylex';
import { COMPONENT_VERSION_ID } from '../config-provider';
import { Highlighter } from '../shared';
import {
  UD_SELECT_OPTION_GROUP_ID,
  UD_SELECT_OPTION_ID,
  useSelectListGroupCssScope,
  useSelectListItemCssScope,
} from './common';
import type {
  InternalSelectOptionGroupProps,
  InternalSelectOptionProps,
  SelectOptionGroupProps,
  SelectOptionProps,
} from './interface';
import { SelectContext } from './context/SelectContext';
import type { SelectValue } from './internal.interface';
import { cssTypographyHeadline } from '../stylex';
import { Text } from '../text/Text';

/**
 * For users
 */
export const SelectOption = <V extends SelectValue>(props: SelectOptionProps<V>) => <></>;
Object.assign(SelectOption, {
  [UD_SELECT_OPTION_ID]: true,
  [COMPONENT_VERSION_ID]: process.env.__VERSION__,
});

/**
 * For internal use
 */
export const InternalSelectOption = forwardRef(
  <V extends SelectValue>(
    {
      value,
      label,
      labelRenderer,
      disabled = false,
      menuItemSelectedIcon,
      title,
      active,
      className,
      // omit
      children,
      ...restProps
    }: PropsWithChildren<InternalSelectOptionProps<V>>,
    ref: Ref<HTMLDivElement>,
  ) => {
    const cssScope = useSelectListItemCssScope();
    const { valueSet, toggle, inputValue, textOverflow } = React.useContext(SelectContext);
    function renderHighligherNode(_value: string) {
      return (
        <Highlighter
          searchWords={[inputValue]}
          textToHighlight={_value}
          markRenderer={text => <span className={clsx(!disabled && cssScope('__content-highlight'))}>{text}</span>}
        />
      );
    }

    const selected = valueSet.has(value);
    let labelNode;
    if (labelRenderer) {
      labelNode = labelRenderer(inputValue);
    } else if (label) {
      labelNode = label;
      if (typeof label === 'string') {
        labelNode = renderHighligherNode(label);
      }
    } else {
      labelNode = renderHighligherNode(String(value));
    }
    return (
      <div
        ref={ref}
        className={clsx(
          className,
          cssScope(),
          selected && cssScope('-selected'),
          active && cssScope('-active'),
          disabled && cssScope('-disabled'),
        )}
        onClick={() => {
          toggle(value);
        }}
        title={title}
        {...restProps}
      >
        <Text
          as="div"
          className={cssScope('__content')}
          lineClamp={textOverflow ? 1 : undefined}
          sx={[stylex.create({ minWidth: 0 })]}
        >
          {labelNode}
        </Text>
        <div className={cssScope('__check')}>{selected ? menuItemSelectedIcon : null}</div>
      </div>
    );
  },
);

/**
 * For users
 */
export const SelectOptionGroup = <V extends SelectValue>(props: SelectOptionGroupProps<V>) => <></>;
Object.assign(SelectOptionGroup, {
  [UD_SELECT_OPTION_GROUP_ID]: true,
  [COMPONENT_VERSION_ID]: process.env.__VERSION__,
});

export const InternalSelectOptionGroup = forwardRef(
  <V extends SelectValue>(
    { isFirst, label, children }: PropsWithChildren<InternalSelectOptionGroupProps<V>>,
    ref: Ref<HTMLDivElement>,
  ) => {
    const cssScope = useSelectListGroupCssScope();
    return (
      <div ref={ref} className={clsx(cssScope(), isFirst && cssScope('-first'))}>
        <div className={clsx(cssScope('__title'), stylex(cssTypographyHeadline))}>{label}</div>
        {children}
      </div>
    );
  },
);
