// @universe-design/react-ai/lib-commonjs/star-thinking-animation/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StarThinkingAnimation = void 0;
var _u874iyf = _interopRequireDefault(require("./../.stylex/u874iyf"));
var _uzqxen = _interopRequireDefault(require("./../.stylex/uzqxen4"));
var _uxirdmd = _interopRequireDefault(require("./../.stylex/uxirdmd"));
var _uupcts = _interopRequireDefault(require("./../.stylex/uupcts6"));
var _uivn5xe = _interopRequireDefault(require("./../.stylex/uivn5xe"));
var _u18zhii = _interopRequireDefault(require("./../.stylex/u18zhii8"));
var _u14enhrg = _interopRequireDefault(require("./../.stylex/u14enhrg"));
var _u1u1p7mj = _interopRequireDefault(require("./../.stylex/u1u1p7mj"));
var _u1jmni = _interopRequireDefault(require("./../.stylex/u1jmni06"));
var _ufibavv = _interopRequireDefault(require("./../.stylex/ufibavv"));
var _uqut2en = _interopRequireDefault(require("./../.stylex/uqut2en"));
var _u1k0zs9m = _interopRequireDefault(require("./../.stylex/u1k0zs9m"));
var _u73r7in = _interopRequireDefault(require("./../.stylex/u73r7in"));
var _u8wgnjj = _interopRequireDefault(require("./../.stylex/u8wgnjj"));
var _u5jxv = _interopRequireDefault(require("./../.stylex/u5jxv71"));
var _u1a7olnx = _interopRequireDefault(require("./../.stylex/u1a7olnx"));
var _u4hyup = _interopRequireDefault(require("./../.stylex/u4hyup3"));
var _u1ma4v2k = _interopRequireDefault(require("./../.stylex/u1ma4v2k"));
var _react = _interopRequireDefault(require("react"));
var _react2 = require("@universe-design/stylex/react");
var _MeetingAiOutlined = _interopRequireDefault(require("@universe-design/icons-react/MeetingAiOutlined"));
var _MeetingAiFilled = _interopRequireDefault(require("@universe-design/icons-react/MeetingAiFilled"));
var _MeetingAiTestFilled = _interopRequireDefault(require("@universe-design/icons-react/MeetingAiTestFilled"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
require("./index.less");
//.u18ujvr1{top:85.2%;}
//.ub5ljux{animation:ai__star-thinking__sub-lb__animation 2.15s cubic-bezier(0.45, 0, 0.55, 1) infinite;}
//.uwodgp6{left:84.4%;}
//.u1grlrm6{animation:ai__star-thinking__sub-rt__animation 2.15s cubic-bezier(0.45, 0, 0.55, 1) infinite;}
//.ukpg2sn{left:15.6%;}
//.u1awxekb{top:14.9%;}
//.uczczkq{animation:ai__star-thinking__sub-lt__animation 2.15s cubic-bezier(0.45, 0, 0.55, 1) infinite;}
//.u1euv43f{position:absolute;}
//.u1ktbn1t{left:50%;}
//.u1i1t8d1{top:50%;}
//.uou3m91{font-size:var(--svg-size);}
//.u17ctas6{animation:ai__star-thinking__main__animation 2.15s cubic-bezier(0.37, 0, 0.63, 1) infinite;}
//.uparq7u{--font-size:1em;}
//.u14t3ns0{display:inline-block;}
//.u10pi13n{position:relative;}
//.u1cyak3s{width:var(--svg-size);}
//.u18fh3uv{height:var(--svg-size);}
//.u15xaump{vertical-align:calc(0.36 * var(--font-size) - 0.5 * var(--svg-size));}
/* eslint-disable react/no-unknown-property */
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const useStarThinkingAnimationStyles = (0, _react2.makeStylexStyles)(_ref => {
  let {
    prefix
  } = _ref;
  return {
    root: {
      "props": {
        "B3yy3ex": _u73r7in.default,
        "mc9l5x": _u8wgnjj.default,
        "qhf8xq": _u5jxv.default,
        "a9b677": _u1a7olnx.default,
        "Bqenvij": _u4hyup.default,
        "ha4doy": _u1ma4v2k.default
      },
      "alias": `${prefix}__star-thinking`
    },
    main: {
      "props": {
        "qhf8xq": _u1u1p7mj.default,
        "oyh7mz": _u1jmni.default,
        "Bhzewxz": _ufibavv.default,
        "Be2twd7": _uqut2en.default,
        "Ba009jy": _u1k0zs9m.default
      },
      "alias": `${prefix}__star-thinking__main`
    },
    sub: {
      base: {
        "props": {
          "qhf8xq": _u1u1p7mj.default,
          "Be2twd7": _uqut2en.default
        },
        "alias": `${prefix}__star-thinking__sub`
      },
      lt: {
        "props": {
          "oyh7mz": _uivn5xe.default,
          "Bhzewxz": _u18zhii.default,
          "Ba009jy": _u14enhrg.default
        },
        "alias": `${prefix}__star-thinking__sub-lt`
      },
      rt: {
        "props": {
          "oyh7mz": _uxirdmd.default,
          "Bhzewxz": _u18zhii.default,
          "Ba009jy": _uupcts.default
        },
        "alias": `${prefix}__star-thinking__sub-rt`
      },
      lb: {
        "props": {
          "oyh7mz": _uivn5xe.default,
          "Bhzewxz": _u874iyf.default,
          "Ba009jy": _uzqxen.default
        },
        "alias": `${prefix}__star-thinking__sub-lb`
      }
    }
  };
});

/**
 * 星型思考动画图标组件入参
 *
 * @public
 */

/**
 * 星型思考动画图标
 *
 * @public
 */
const StarThinkingAnimation = props => {
  const {
    type = 'outlined',
    size = 24,
    style,
    className,
    sx
  } = props;
  const styles = useStarThinkingAnimationStyles({
    prefix: 'ai'
  });
  return (0, _react2.jsx)("div", {
    sx: [styles.root, sx],
    style: {
      '--svg-size': typeof size === 'number' ? `${size}px` : size,
      ...style
    },
    className: className
  }, type === 'filled' ? (0, _react2.jsx)(_MeetingAiFilled.default, {
    className: (0, _stylex.default)(styles.main)
  }) : (0, _react2.jsx)(_MeetingAiOutlined.default, {
    className: (0, _stylex.default)(styles.main)
  }), (0, _react2.jsx)(_MeetingAiTestFilled.default, {
    className: (0, _stylex.default)(styles.sub.base, styles.sub.lt)
  }), (0, _react2.jsx)(_MeetingAiTestFilled.default, {
    className: (0, _stylex.default)(styles.sub.base, styles.sub.rt)
  }), (0, _react2.jsx)(_MeetingAiTestFilled.default, {
    className: (0, _stylex.default)(styles.sub.base, styles.sub.lb)
  }));
};
exports.StarThinkingAnimation = StarThinkingAnimation;
