// @universe-design/react/lib/shared/components/highlighter/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Highlighter = exports.HighlightProvider = exports.HighlightConsumer = void 0;
var _react = _interopRequireDefault(require("react"));
var _highlightWordsCore = require("highlight-words-core");
var _useContextSelector = require("use-context-selector");
const HightLightContext = (0, _useContextSelector.createContext)({
  disabled: false
});
const defaultTextRenderer = text => /*#__PURE__*/_react.default.createElement("span", null, text);
const defaultMarkRenderer = text => /*#__PURE__*/_react.default.createElement("mark", null, text);
const Highlighter = function (_ref) {
  let {
    autoEscape = true,
    caseSensitive = false,
    findChunks,
    sanitize,
    searchWords,
    textToHighlight,
    markRenderer = defaultMarkRenderer,
    textRenderer = defaultTextRenderer
  } = _ref;
  const chunks = (0, _highlightWordsCore.findAll)({
    autoEscape,
    caseSensitive,
    findChunks,
    sanitize,
    searchWords,
    textToHighlight
  });
  return /*#__PURE__*/_react.default.createElement("span", null, chunks.map((chunk, index) => {
    const text = textToHighlight.substr(chunk.start, chunk.end - chunk.start);
    if (chunk.highlight) {
      const markNode = markRenderer(text);
      return /*#__PURE__*/_react.default.isValidElement(markNode) ? /*#__PURE__*/_react.default.cloneElement(markNode, {
        key: index
      }) : /*#__PURE__*/_react.default.cloneElement(defaultMarkRenderer(text), {
        key: index
      });
    }
    const textNode = textRenderer(text);
    return /*#__PURE__*/_react.default.isValidElement(textNode) ? /*#__PURE__*/_react.default.cloneElement(textNode, {
      key: index
    }) : /*#__PURE__*/_react.default.cloneElement(defaultTextRenderer(text), {
      key: index
    });
  }));
};
exports.Highlighter = Highlighter;
const HighlightProvider = exports.HighlightProvider = HightLightContext.Provider;
if (process.env.NODE_ENV !== "production") {
  HighlightProvider.displayName = 'HighlightProvider';
}
const HighlightConsumer = _ref2 => {
  let {
    markRenderer,
    children
  } = _ref2;
  const {
    searchWords,
    disabled
  } = (0, _useContextSelector.useContext)(HightLightContext);
  if (typeof children === 'string' && searchWords && searchWords.length && !disabled) {
    return /*#__PURE__*/_react.default.createElement(Highlighter, {
      textToHighlight: children,
      searchWords: searchWords,
      markRenderer: markRenderer
    });
  }
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, children);
};
exports.HighlightConsumer = HighlightConsumer;
if (process.env.NODE_ENV !== "production") {
  HighlightConsumer.displayName = 'HighlightConsumer';
}
