// @universe-design/react/src/timeline/TimelineItem.tsx
import React from 'react';
import clsx from 'clsx';
import TimeFilled from '@universe-design/icons-react/TimeFilled';
import TimeOutlined from '@universe-design/icons-react/TimeOutlined';
import CloseFilled from '@universe-design/icons-react/CloseFilled';
import WarningFilled from '@universe-design/icons-react/WarningFilled';
import YesFilled from '@universe-design/icons-react/YesFilled';
import YesOutlined from '@universe-design/icons-react/YesOutlined';
import { useCssScope } from '../config-provider/hooks';
import type { TimelineItemProps, TimelineStatusType } from './types';
import stylex from '@universe-design/stylex';
import { cssTypographyBody0, cssTypographyBody2 } from '../stylex';
import { SpinLoadingSvg } from '../_assets/svgs';

const useTimelineItemCssScope = () => useCssScope('timeline-item');

const iconStatusMap: Record<TimelineStatusType, React.ReactNode> = {
  success: <YesFilled />,
  warning: <WarningFilled />,
  error: <CloseFilled />,
  done: <YesOutlined />,
  current: <TimeFilled />,
  notarrived: <TimeOutlined />,
  loading: <SpinLoadingSvg />,
};

const TimelineItem: React.FC<TimelineItemProps> = props => {
  const {
    position,
    labelPosition = 'relative',
    className,
    title,
    label,
    description,
    status,
    dotType = 'dot',
    dot = dotType === 'icon' && status && iconStatusMap[status],
    ...reset
  } = props;
  const itemCssScope = useTimelineItemCssScope();
  const labelEl = <div className={clsx(itemCssScope('__label'), stylex(cssTypographyBody2))}>{label}</div>;
  const same = labelPosition === 'same';

  return (
    <li
      {...reset}
      className={clsx(itemCssScope(), className, {
        [itemCssScope('--custom-dot')]: dot,
        [itemCssScope(`--position-${position}`)]: position,
        [itemCssScope(`--status-${status}`)]: status,
      })}
    >
      <div className={itemCssScope('__oppsite-content')}>{!same && label && labelEl}</div>
      <div className={itemCssScope('__seperator')}>
        {dot ? <div className={itemCssScope('__dot--custom')}>{dot}</div> : <div className={itemCssScope('__dot')} />}
        <div className={itemCssScope('__tail')} />
      </div>
      <div className={itemCssScope('__content')}>
        {title && <div className={clsx(itemCssScope('__title'), stylex(cssTypographyBody0))}>{title}</div>}
        {same && label && labelEl}
        {description && <div className={clsx(itemCssScope('__des'), stylex(cssTypographyBody2))}>{description}</div>}
      </div>
    </li>
  );
};

if (__DEV__) {
  TimelineItem.displayName = 'Timeline.Item';
}

export default TimelineItem;
