// @universe-design/react/src/timeline/types.ts
import type { CSSProperties } from 'react';
import type React from 'react';
import type { DefaultSizeType } from '../util';
import type TimelineItem from './TimelineItem';

export type TimelineModeType = 'left' | 'right' | 'alternate';
export type TimelinePositionType = 'left' | 'right';
export type TimelineSpaceType = DefaultSizeType;
export type TimelineLabelPosition = 'same' | 'relative';
export type TimelineDotType = 'dot' | 'icon';
export type TimelineStatusType = 'success' | 'warning' | 'error' | 'done' | 'current' | 'notarrived' | 'loading';

/**
 * @public
 */
export interface TimelineProps extends React.PropsWithChildren<unknown> {
  /**
   * `className` 属性透传
   */
  className?: string;
  /**
   * `style` 属性透传
   */
  style?: CSSProperties;
  /**
   * 设置时间轴和内容的相对位置
   * @defaultValue `left`
   */
  mode?: TimelineModeType;
  /**
   * 节点是否反序排列
   * @defaultValue false
   */
  reverse?: boolean;
  /**
   * 最后一个节点是否展示孤立的轴线
   * @defaultValue false
   */
  hangingTail?: boolean;
  /**
   * 是否在末尾添加一个幽灵节点
   * @defaultValue false
   */
  pending?: boolean | React.ReactNode;
  /**
   * 自定义幽灵节点的图标
   * @defaultValue `<ChatLoadingOutlined />`
   */
  pendingDot?: React.ReactNode;
  /**
   * 设置节点之间的间距
   * @defaultValue `medium`
   */
  space?: TimelineSpaceType;
  /**
   * 标签文本定位关系，设置为 `same` 时始终跟随 title 进行定位
   * @defaultValue `relative`
   */
  labelPosition?: TimelineLabelPosition;
}

export interface TimelineType extends React.FC<TimelineProps> {
  Item: typeof TimelineItem;
}

/**
 * <Timeline.Item> 的属性
 * @public
 */
export interface TimelineItemProps {
  /**
   * 自定义时间轴节点。当 `dotType` 为 `icon` 时，默认会根据 `status` 来切换节点图标。
   */
  dot?: React.ReactNode;
  /**
   * 设置标题
   */
  title?: React.ReactNode;
  /**
   * 设置标签
   */
  label?: React.ReactNode;
  /**
   * 设置辅助描述信息
   */
  description?: React.ReactNode;
  /**
   * 时间轴 `mode` 为 `alternate` 时，节点的位置
   */
  position?: TimelinePositionType;
  /**
   * 标签文本定位关系，设置为 `same` 时始终跟随 title 进行定位
   * @defaultValue `relative`
   */
  labelPosition?: TimelineLabelPosition;
  /**
   * `className` 属性透传
   */
  className?: string;
  /**
   * `style` 属性透传
   */
  style?: CSSProperties;
  /**
   * 节点状态
   * @defaultValue `done`
   */
  status?: TimelineStatusType;
  /**
   * 时间轴节点类型
   * @defaultValue `dot`
   */
  dotType?: TimelineDotType;
}
