// @universe-design/react/src/pro/organization-name-tag/stories/site.stories.tsx
// autogen by ../../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Basic from '../demo/0.basic';
import C1Size from '../demo/1.size';
import C2Overflow from '../demo/2.overflow';
import C3DynamicWidth from '../demo/3.dynamic-width';
if (!(C0Basic as Story).storyName) {
  (C0Basic as Story).storyName = '0. 基础用法';
  (C0Basic as Story).parameters = {
    docs: {
      description: {
        story:
          '通用蓝色标签，无 icon。业务不传入 `children` 时，触发兜底策略，固定展示「外部」，同时悬停支持提示相关信息。\n\n',
      },
    },
  };
}
if (!(C1Size as Story).storyName) {
  (C1Size as Story).storyName = '1. 不同尺寸';
  (C1Size as Story).parameters = {
    docs: {
      description: {
        story: '支持 `Tag` 的 4 种尺寸，默认值为 `extra-small`\n\n',
      },
    },
  };
}
if (!(C2Overflow as Story).storyName) {
  (C2Overflow as Story).storyName = '2. 溢出弹层';
  (C2Overflow as Story).parameters = {
    docs: {
      description: {
        story: '展示为缺省「外部」标签时，溢出弹层功能会关闭，展示为企业名称时，弹层内容为公司全称。\n\n',
      },
    },
  };
}
if (!(C3DynamicWidth as Story).storyName) {
  (C3DynamicWidth as Story).storyName = '3. 动态调整「最小缺省宽度」';
  (C3DynamicWidth as Story).parameters = {
    docs: {
      description: {
        story:
          '浏览器默认会把 `单字符 + ...` 作为文本的「最小缺省宽度」，该示例提供了一种动态调整「最小缺省宽度」的方法。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/OrganizationNameTag/官网示例',
  tags: ['autodocs'],
};

export { C0Basic, C1Size, C2Overflow, C3DynamicWidth };
