// @universe-design/styles/cjs/universe-design/index.light.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.udtoken_colorpicker_orange = exports.udtoken_colorpicker_neutral = exports.udtoken_colorpicker_lime = exports.udtoken_colorpicker_indigo = exports.udtoken_colorpicker_green = exports.udtoken_colorpicker_carmine = exports.udtoken_colorpicker_blue = exports.udtoken_chatsender_stopgeneration_bg2_normal = exports.udtoken_chatsender_stopgeneration_bg1_pressed = exports.udtoken_chatsender_stopgeneration_bg1_normal = exports.udtoken_chatsender_stopgeneration_bg1_hover = exports.udtoken_btn_text_bg_pri_pressed = exports.udtoken_btn_text_bg_pri_hover = exports.udtoken_btn_text_bg_pri_focus = exports.udtoken_btn_text_bg_neutral_pressed = exports.udtoken_btn_text_bg_neutral_hover = exports.udtoken_btn_text_bg_neutral_focus = exports.udtoken_btn_text_bg_danger_pressed = exports.udtoken_btn_text_bg_danger_hover = exports.udtoken_btn_text_bg_danger_focus = exports.udtoken_btn_selected_text_success = exports.udtoken_btn_selected_text_danger = exports.udtoken_btn_selected_border_neutral = exports.udtoken_btn_selected_bg_success_press = exports.udtoken_btn_selected_bg_success_normal = exports.udtoken_btn_selected_bg_success_hover = exports.udtoken_btn_selected_bg_neutral_press = exports.udtoken_btn_selected_bg_neutral_normal = exports.udtoken_btn_selected_bg_neutral_hover = exports.udtoken_btn_selected_bg_danger_press = exports.udtoken_btn_selected_bg_danger_normal = exports.udtoken_btn_selected_bg_danger_hover = exports.udtoken_btn_se_bg_success_pressed = exports.udtoken_btn_se_bg_success_hover = exports.udtoken_btn_se_bg_success_focus = exports.udtoken_btn_se_bg_pri_pressed = exports.udtoken_btn_se_bg_pri_hover = exports.udtoken_btn_se_bg_pri_focus = exports.udtoken_btn_se_bg_neutral_pressed = exports.udtoken_btn_se_bg_neutral_hover = exports.udtoken_btn_se_bg_neutral_focus = exports.udtoken_btn_se_bg_danger_pressed = exports.udtoken_btn_se_bg_danger_hover = exports.udtoken_btn_se_bg_danger_focus = exports.udtoken_btn_pri_text_disabled = exports.udtoken_btn_ghost_bg = exports.udtoken_block_view_inline_bg_self = exports.udtoken_block_view_inline_bg_other = exports.udtoken_block_view_inline_bg_nopermission = exports.udtoken_AI_cot_shinytext_mask = void 0;
exports.udtoken_navigation_bar_tab_bg_black = exports.udtoken_navigation_bar_resize_handle = exports.udtoken_message_card_text_yellow = exports.udtoken_message_card_text_wathet = exports.udtoken_message_card_text_violet = exports.udtoken_message_card_text_turquoise = exports.udtoken_message_card_text_red = exports.udtoken_message_card_text_purple = exports.udtoken_message_card_text_orange = exports.udtoken_message_card_text_neutral = exports.udtoken_message_card_text_lime = exports.udtoken_message_card_text_indigo = exports.udtoken_message_card_text_green = exports.udtoken_message_card_text_carmine = exports.udtoken_message_card_text_blue = exports.udtoken_message_card_bg_yellow = exports.udtoken_message_card_bg_wathet = exports.udtoken_message_card_bg_violet = exports.udtoken_message_card_bg_turquoise = exports.udtoken_message_card_bg_red = exports.udtoken_message_card_bg_purple = exports.udtoken_message_card_bg_orange = exports.udtoken_message_card_bg_neutral = exports.udtoken_message_card_bg_mask_special = exports.udtoken_message_card_bg_mask_general = exports.udtoken_message_card_bg_lime = exports.udtoken_message_card_bg_light_neutral = exports.udtoken_message_card_bg_indigo = exports.udtoken_message_card_bg_green = exports.udtoken_message_card_bg_carmine = exports.udtoken_message_card_bg_blue = exports.udtoken_input_bg_disabled = exports.udtoken_imageview_tooltips_bg = exports.udtoken_imageview_toolbox_bg = exports.udtoken_imageview_icondisabled = exports.udtoken_imageview_bg_mask = exports.udtoken_dropdown_button_divider = exports.udtoken_component_text_disabled_loading = exports.udtoken_component_outlined_bg = exports.udtoken_component_bg_top_mask = exports.udtoken_component_bg_right_mask = exports.udtoken_component_bg_left_mask = exports.udtoken_component_bg_bottom_mask = exports.udtoken_colorpicker_yellow = exports.udtoken_colorpicker_wathet = exports.udtoken_colorpicker_violet = exports.udtoken_colorpicker_turquoise = exports.udtoken_colorpicker_sunflower = exports.udtoken_colorpicker_red = exports.udtoken_colorpicker_purple = void 0;
exports.udtoken_tag_bg_purple = exports.udtoken_tag_bg_orange_solid = exports.udtoken_tag_bg_orange_pressed = exports.udtoken_tag_bg_orange_hover = exports.udtoken_tag_bg_orange = exports.udtoken_tag_bg_lime_solid = exports.udtoken_tag_bg_lime_hover = exports.udtoken_tag_bg_lime = exports.udtoken_tag_bg_indigo_solid = exports.udtoken_tag_bg_indigo_hover = exports.udtoken_tag_bg_indigo = exports.udtoken_tag_bg_green_solid = exports.udtoken_tag_bg_green_hover = exports.udtoken_tag_bg_green = exports.udtoken_tag_bg_carmine_solid = exports.udtoken_tag_bg_carmine_hover = exports.udtoken_tag_bg_carmine = exports.udtoken_tag_bg_blue_solid = exports.udtoken_tag_bg_blue_pressed = exports.udtoken_tag_bg_blue_hover = exports.udtoken_tag_bg_blue_default = exports.udtoken_tag_bg_blue = exports.udtoken_tag_AI_trial_bg_text_normal = exports.udtoken_tag_AI_trial_bg_normal = exports.udtoken_tabs_view_bg_hover = exports.udtoken_tabs_line_border_card = exports.udtoken_table_bg_selected = exports.udtoken_table_bg_pressed = exports.udtoken_table_bg_hover = exports.udtoken_table_bg_head = exports.udtoken_table_bg_grey = exports.udtoken_table_bg_expand = exports.udtoken_tab_se_bg_unselected = exports.udtoken_tab_pri_text = exports.udtoken_tab_pri_bg = exports.udtoken_switch_handle_disabled = exports.udtoken_steps_bg_hover = exports.udtoken_sliding_block_bg_disabled_loading = exports.udtoken_skeleton_fg = exports.udtoken_skeleton_bg = exports.udtoken_side_navigation_fill_selected = exports.udtoken_side_navigation_fill_pressed = exports.udtoken_side_navigation_fill_hover = exports.udtoken_side_navigation_fill_focus = exports.udtoken_reaction_bg_grey_float = exports.udtoken_reaction_bg_grey = exports.udtoken_quote_bar_bg = exports.udtoken_progress_bg = exports.udtoken_navigation_bar_tab_bg_gray = exports.udtoken_navigation_bar_tab_bg_blue = void 0;
exports.udtoken_tag_text_lime = exports.udtoken_tag_text_indigo = exports.udtoken_tag_text_green = exports.udtoken_tag_text_carmine = exports.udtoken_tag_text_blue = exports.udtoken_tag_state_red_light = exports.udtoken_tag_state_purple_light = exports.udtoken_tag_state_orange_light = exports.udtoken_tag_state_grey_light = exports.udtoken_tag_state_green_light = exports.udtoken_tag_state_blue_light = exports.udtoken_tag_neutral_text_solid = exports.udtoken_tag_neutral_text_normal = exports.udtoken_tag_neutral_text_inverse_opacity = exports.udtoken_tag_neutral_text_inverse = exports.udtoken_tag_neutral_close_pressed = exports.udtoken_tag_neutral_close_normal = exports.udtoken_tag_neutral_close_hover = exports.udtoken_tag_neutral_close_disabled = exports.udtoken_tag_neutral_bg_solid = exports.udtoken_tag_neutral_bg_normal_pressed = exports.udtoken_tag_neutral_bg_normal_hover = exports.udtoken_tag_neutral_bg_normal = exports.udtoken_tag_neutral_bg_inverse_opacity = exports.udtoken_tag_neutral_bg_inverse = exports.udtoken_tag_close_orange_pressed = exports.udtoken_tag_close_orange_normal = exports.udtoken_tag_close_orange_hover = exports.udtoken_tag_close_orange_disabled = exports.udtoken_tag_close_blue_pressed = exports.udtoken_tag_close_blue_normal = exports.udtoken_tag_close_blue_hover = exports.udtoken_tag_close_blue_disabled = exports.udtoken_tag_bg_yellow_solid = exports.udtoken_tag_bg_yellow_hover = exports.udtoken_tag_bg_yellow = exports.udtoken_tag_bg_wathet_solid = exports.udtoken_tag_bg_wathet_hover = exports.udtoken_tag_bg_wathet = exports.udtoken_tag_bg_violet_solid = exports.udtoken_tag_bg_violet_hover = exports.udtoken_tag_bg_violet = exports.udtoken_tag_bg_turquoise_solid = exports.udtoken_tag_bg_turquoise_hover = exports.udtoken_tag_bg_turquoise = exports.udtoken_tag_bg_red_solid = exports.udtoken_tag_bg_red_hover = exports.udtoken_tag_bg_red = exports.udtoken_tag_bg_purple_solid = exports.udtoken_tag_bg_purple_hover = void 0;
exports.udtoken_upload_mask_img = exports.udtoken_upload_bg_error = exports.udtoken_tag_text_yellow = exports.udtoken_tag_text_wathet = exports.udtoken_tag_text_violet = exports.udtoken_tag_text_turquoise = exports.udtoken_tag_text_s_yellow = exports.udtoken_tag_text_s_wathet = exports.udtoken_tag_text_s_violet = exports.udtoken_tag_text_s_turquoise = exports.udtoken_tag_text_s_red = exports.udtoken_tag_text_s_purple = exports.udtoken_tag_text_s_orange = exports.udtoken_tag_text_s_lime = exports.udtoken_tag_text_s_indigo = exports.udtoken_tag_text_s_green = exports.udtoken_tag_text_s_carmine = exports.udtoken_tag_text_s_blue = exports.udtoken_tag_text_red = exports.udtoken_tag_text_purple = exports.udtoken_tag_text_orange = void 0;
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
exports.udtoken_AI_cot_shinytext_mask = 'rgba(255, 255, 255, 0.6)';
exports.udtoken_block_view_inline_bg_nopermission = 'rgba(31, 35, 41, 0.1)';
exports.udtoken_block_view_inline_bg_other = 'rgba(148, 180, 255, 0.18)';
exports.udtoken_block_view_inline_bg_self = '#1456f0';
exports.udtoken_btn_ghost_bg = 'rgba(255, 255, 255, 0)';
exports.udtoken_btn_pri_text_disabled = '#ffffff';
exports.udtoken_btn_se_bg_danger_focus = '#fee3e2';
exports.udtoken_btn_se_bg_danger_hover = '#fee3e2';
exports.udtoken_btn_se_bg_danger_pressed = '#fdc6c4';
exports.udtoken_btn_se_bg_neutral_focus = '#f2f3f5';
exports.udtoken_btn_se_bg_neutral_hover = '#f2f3f5';
exports.udtoken_btn_se_bg_neutral_pressed = '#dee0e3';
exports.udtoken_btn_se_bg_pri_focus = '#e0e9ff';
exports.udtoken_btn_se_bg_pri_hover = '#e0e9ff';
exports.udtoken_btn_se_bg_pri_pressed = '#c2d4ff';
exports.udtoken_btn_se_bg_success_focus = '#d0f5ce';
exports.udtoken_btn_se_bg_success_hover = '#d0f5ce';
exports.udtoken_btn_se_bg_success_pressed = '#95e599';
exports.udtoken_btn_selected_bg_danger_hover = '#fee3e2';
exports.udtoken_btn_selected_bg_danger_normal = '#fef0f0';
exports.udtoken_btn_selected_bg_danger_press = '#fdc6c4';
exports.udtoken_btn_selected_bg_neutral_hover = 'rgba(31, 35, 41, 0.1)';
exports.udtoken_btn_selected_bg_neutral_normal = 'rgba(31, 35, 41, 0.06)';
exports.udtoken_btn_selected_bg_neutral_press = 'rgba(31, 35, 41, 0.14)';
exports.udtoken_btn_selected_bg_success_hover = '#d0f5ce';
exports.udtoken_btn_selected_bg_success_normal = '#e4fae1';
exports.udtoken_btn_selected_bg_success_press = '#95e599';
exports.udtoken_btn_selected_border_neutral = '#bbbfc4';
exports.udtoken_btn_selected_text_danger = '#c02a26';
exports.udtoken_btn_selected_text_success = '#258832';
exports.udtoken_btn_text_bg_danger_focus = 'rgba(226, 46, 40, 0.1)';
exports.udtoken_btn_text_bg_danger_hover = 'rgba(226, 46, 40, 0.1)';
exports.udtoken_btn_text_bg_danger_pressed = 'rgba(226, 46, 40, 0.2)';
exports.udtoken_btn_text_bg_neutral_focus = 'rgba(31, 35, 41, 0.1)';
exports.udtoken_btn_text_bg_neutral_hover = 'rgba(31, 35, 41, 0.1)';
exports.udtoken_btn_text_bg_neutral_pressed = 'rgba(31, 35, 41, 0.2)';
exports.udtoken_btn_text_bg_pri_focus = 'rgba(20, 86, 240, 0.1)';
exports.udtoken_btn_text_bg_pri_hover = 'rgba(20, 86, 240, 0.1)';
exports.udtoken_btn_text_bg_pri_pressed = 'rgba(20, 86, 240, 0.2)';
exports.udtoken_chatsender_stopgeneration_bg1_hover = 'rgba(31, 35, 41, 0.15)';
exports.udtoken_chatsender_stopgeneration_bg1_normal = 'rgba(31, 35, 41, 0.1)';
exports.udtoken_chatsender_stopgeneration_bg1_pressed = 'rgba(31, 35, 41, 0.2)';
exports.udtoken_chatsender_stopgeneration_bg2_normal = '#1f2329';
exports.udtoken_colorpicker_blue = '#336df4';
exports.udtoken_colorpicker_carmine = '#df58a5';
exports.udtoken_colorpicker_green = '#35bd4b';
exports.udtoken_colorpicker_indigo = '#4752e6';
exports.udtoken_colorpicker_lime = '#a2c10b';
exports.udtoken_colorpicker_neutral = '#8f959e';
exports.udtoken_colorpicker_orange = '#ff811a';
exports.udtoken_colorpicker_purple = '#7a35f0';
exports.udtoken_colorpicker_red = '#f54a45';
exports.udtoken_colorpicker_sunflower = '#fff67a';
exports.udtoken_colorpicker_turquoise = '#33d6c0';
exports.udtoken_colorpicker_violet = '#bf3dbf';
exports.udtoken_colorpicker_wathet = '#3ec3f7';
exports.udtoken_colorpicker_yellow = '#ffc60a';
exports.udtoken_component_bg_bottom_mask = 'linear-gradient(to bottom, rgba(255, 255, 255, 0) 0%, #ffffff 100%)';
exports.udtoken_component_bg_left_mask = 'linear-gradient(to left, rgba(255, 255, 255, 0) 0%, #ffffff 100%)';
exports.udtoken_component_bg_right_mask = 'linear-gradient(to right, rgba(255, 255, 255, 0) 0%, #ffffff 100%)';
exports.udtoken_component_bg_top_mask = 'linear-gradient(to top, rgba(255, 255, 255, 0) 0%, #ffffff 100%)';
exports.udtoken_component_outlined_bg = '#ffffff';
exports.udtoken_component_text_disabled_loading = '#8f959e';
exports.udtoken_dropdown_button_divider = 'rgba(255, 255, 255, 0.5)';
exports.udtoken_imageview_bg_mask = 'rgba(0, 0, 0, 0.8)';
exports.udtoken_imageview_icondisabled = 'rgba(255, 255, 255, 0.6)';
exports.udtoken_imageview_toolbox_bg = 'rgba(0, 0, 0, 0.8)';
exports.udtoken_imageview_tooltips_bg = '#505050';
exports.udtoken_input_bg_disabled = '#eff0f1';
exports.udtoken_message_card_bg_blue = '#e0e9ff';
exports.udtoken_message_card_bg_carmine = '#fee2f2';
exports.udtoken_message_card_bg_green = '#d0f5ce';
exports.udtoken_message_card_bg_indigo = '#e9eafb';
exports.udtoken_message_card_bg_light_neutral = '#f2f3f5';
exports.udtoken_message_card_bg_lime = '#e3f0a3';
exports.udtoken_message_card_bg_mask_general = 'rgba(255, 255, 255, 0.5)';
exports.udtoken_message_card_bg_mask_special = 'rgba(255, 255, 255, 0.2)';
exports.udtoken_message_card_bg_neutral = '#8f959e';
exports.udtoken_message_card_bg_orange = '#fee7cd';
exports.udtoken_message_card_bg_purple = '#efe6fe';
exports.udtoken_message_card_bg_red = '#fee3e2';
exports.udtoken_message_card_bg_turquoise = '#c4f2ec';
exports.udtoken_message_card_bg_violet = '#f9e2f9';
exports.udtoken_message_card_bg_wathet = '#caeffc';
exports.udtoken_message_card_bg_yellow = '#faedc2';
exports.udtoken_message_card_text_blue = '#1456f0';
exports.udtoken_message_card_text_carmine = '#b82879';
exports.udtoken_message_card_text_green = '#1a7526';
exports.udtoken_message_card_text_indigo = '#4752e6';
exports.udtoken_message_card_text_lime = '#5c6d08';
exports.udtoken_message_card_text_neutral = '#ffffff';
exports.udtoken_message_card_text_orange = '#c25705';
exports.udtoken_message_card_text_purple = '#7a35f0';
exports.udtoken_message_card_text_red = '#c02a26';
exports.udtoken_message_card_text_turquoise = '#067062';
exports.udtoken_message_card_text_violet = '#a630a6';
exports.udtoken_message_card_text_wathet = '#076a94';
exports.udtoken_message_card_text_yellow = '#ad7a03';
exports.udtoken_navigation_bar_resize_handle = '#4599ff';
exports.udtoken_navigation_bar_tab_bg_black = '#1c2435';
exports.udtoken_navigation_bar_tab_bg_blue = '#1b429c';
exports.udtoken_navigation_bar_tab_bg_gray = '#e5e5e9';
exports.udtoken_progress_bg = '#dee0e3';
exports.udtoken_quote_bar_bg = '#bbbfc4';
exports.udtoken_reaction_bg_grey = 'rgba(31, 35, 41, 0.05)';
exports.udtoken_reaction_bg_grey_float = 'rgba(31, 35, 41, 0.05)';
exports.udtoken_side_navigation_fill_focus = 'rgba(31, 35, 41, 0.12)';
exports.udtoken_side_navigation_fill_hover = 'rgba(31, 35, 41, 0.05)';
exports.udtoken_side_navigation_fill_pressed = 'rgba(31, 35, 41, 0.12)';
exports.udtoken_side_navigation_fill_selected = 'rgba(31, 35, 41, 0.1)';
exports.udtoken_skeleton_bg = 'rgba(31, 35, 41, 0.08)';
exports.udtoken_skeleton_fg = 'rgba(31, 35, 41, 0.05)';
exports.udtoken_sliding_block_bg_disabled_loading = '#ffffff';
exports.udtoken_steps_bg_hover = 'rgba(20, 86, 240, 0.1)';
exports.udtoken_switch_handle_disabled = 'rgba(255, 255, 255, 0.6)';
exports.udtoken_tab_pri_bg = 'rgba(20, 86, 240, 0.2)';
exports.udtoken_tab_pri_text = '#1456f0';
exports.udtoken_tab_se_bg_unselected = 'rgba(31, 35, 41, 0.1)';
exports.udtoken_table_bg_expand = '#f2f3f5';
exports.udtoken_table_bg_grey = '#f5f6f7';
exports.udtoken_table_bg_head = '#f2f3f5';
exports.udtoken_table_bg_hover = '#eff0f1';
exports.udtoken_table_bg_pressed = '#dee0e3';
exports.udtoken_table_bg_selected = '#f0f4ff';
exports.udtoken_tabs_line_border_card = '#dee0e3';
exports.udtoken_tabs_view_bg_hover = 'rgba(31, 35, 41, 0.03)';
exports.udtoken_tag_AI_trial_bg_normal = 'radial-gradient(100.62% 141.42% at 100% 0%, rgba(183, 145, 250, 0.12) 0%, rgba(91, 101, 245, 0.12) 79.5%)';
exports.udtoken_tag_AI_trial_bg_text_normal = 'radial-gradient(58.07% 105.94% at 95% 28.12%, #8d55ed 0%, #4752e6 100%)';
exports.udtoken_tag_bg_blue = 'rgba(20, 86, 240, 0.2)';
exports.udtoken_tag_bg_blue_default = 'rgba(20, 86, 240, 0.12)';
exports.udtoken_tag_bg_blue_hover = 'rgba(20, 86, 240, 0.3)';
exports.udtoken_tag_bg_blue_pressed = 'rgba(20, 86, 240, 0.2)';
exports.udtoken_tag_bg_blue_solid = '#c2d4ff';
exports.udtoken_tag_bg_carmine = 'rgba(204, 57, 140, 0.2)';
exports.udtoken_tag_bg_carmine_hover = 'rgba(204, 57, 140, 0.3)';
exports.udtoken_tag_bg_carmine_solid = '#f8c4e1';
exports.udtoken_tag_bg_green = 'rgba(53, 189, 75, 0.2)';
exports.udtoken_tag_bg_green_hover = 'rgba(53, 189, 75, 0.3)';
exports.udtoken_tag_bg_green_solid = '#95e599';
exports.udtoken_tag_bg_indigo = 'rgba(91, 101, 245, 0.2)';
exports.udtoken_tag_bg_indigo_hover = 'rgba(91, 101, 245, 0.3)';
exports.udtoken_tag_bg_indigo_solid = '#cccff9';
exports.udtoken_tag_bg_lime = 'rgba(145, 173, 0, 0.2)';
exports.udtoken_tag_bg_lime_hover = 'rgba(145, 173, 0, 0.3)';
exports.udtoken_tag_bg_lime_solid = '#c8dd5f';
exports.udtoken_tag_bg_orange = 'rgba(255, 129, 26, 0.15)';
exports.udtoken_tag_bg_orange_hover = 'rgba(255, 129, 26, 0.3)';
exports.udtoken_tag_bg_orange_pressed = 'rgba(255, 129, 26, 0.35)';
exports.udtoken_tag_bg_orange_solid = '#fec48b';
exports.udtoken_tag_bg_purple = 'rgba(141, 85, 237, 0.2)';
exports.udtoken_tag_bg_purple_hover = 'rgba(141, 85, 237, 0.3)';
exports.udtoken_tag_bg_purple_solid = '#dcc9fd';
exports.udtoken_tag_bg_red = 'rgba(245, 74, 69, 0.2)';
exports.udtoken_tag_bg_red_hover = 'rgba(245, 74, 69, 0.3)';
exports.udtoken_tag_bg_red_solid = '#fdc6c4';
exports.udtoken_tag_bg_turquoise = 'rgba(51, 214, 192, 0.2)';
exports.udtoken_tag_bg_turquoise_hover = 'rgba(51, 214, 192, 0.3)';
exports.udtoken_tag_bg_turquoise_solid = '#6fe8d8';
exports.udtoken_tag_bg_violet = 'rgba(191, 61, 191, 0.2)';
exports.udtoken_tag_bg_violet_hover = 'rgba(191, 61, 191, 0.3)';
exports.udtoken_tag_bg_violet_solid = '#f3c4f3';
exports.udtoken_tag_bg_wathet = 'rgba(62, 195, 247, 0.2)';
exports.udtoken_tag_bg_wathet_hover = 'rgba(62, 195, 247, 0.3)';
exports.udtoken_tag_bg_wathet_solid = '#97dcfc';
exports.udtoken_tag_bg_yellow = 'rgba(250, 211, 85, 0.2)';
exports.udtoken_tag_bg_yellow_hover = 'rgba(250, 211, 85, 0.4)';
exports.udtoken_tag_bg_yellow_solid = '#fcdf7e';
exports.udtoken_tag_close_blue_disabled = 'rgba(20, 86, 240, 0.8)';
exports.udtoken_tag_close_blue_hover = '#1456f0';
exports.udtoken_tag_close_blue_normal = 'rgba(20, 86, 240, 0.6)';
exports.udtoken_tag_close_blue_pressed = 'rgba(20, 86, 240, 0.8)';
exports.udtoken_tag_close_orange_disabled = 'rgba(255, 129, 26, 0.3)';
exports.udtoken_tag_close_orange_hover = '#ff811a';
exports.udtoken_tag_close_orange_normal = 'rgba(255, 129, 26, 0.6)';
exports.udtoken_tag_close_orange_pressed = 'rgba(255, 129, 26, 0.8)';
exports.udtoken_tag_neutral_bg_inverse = '#8f959e';
exports.udtoken_tag_neutral_bg_inverse_opacity = 'rgba(31, 35, 41, 0.4)';
exports.udtoken_tag_neutral_bg_normal = 'rgba(31, 35, 41, 0.1)';
exports.udtoken_tag_neutral_bg_normal_hover = 'rgba(31, 35, 41, 0.15)';
exports.udtoken_tag_neutral_bg_normal_pressed = 'rgba(31, 35, 41, 0.2)';
exports.udtoken_tag_neutral_bg_solid = '#eff0f1';
exports.udtoken_tag_neutral_close_disabled = 'rgba(31, 35, 41, 0.3)';
exports.udtoken_tag_neutral_close_hover = '#1f2329';
exports.udtoken_tag_neutral_close_normal = 'rgba(31, 35, 41, 0.6)';
exports.udtoken_tag_neutral_close_pressed = 'rgba(31, 35, 41, 0.8)';
exports.udtoken_tag_neutral_text_inverse = '#ffffff';
exports.udtoken_tag_neutral_text_inverse_opacity = '#ffffff';
exports.udtoken_tag_neutral_text_normal = '#646a73';
exports.udtoken_tag_neutral_text_solid = '#646a73';
exports.udtoken_tag_state_blue_light = '#336df4';
exports.udtoken_tag_state_green_light = '#258832';
exports.udtoken_tag_state_grey_light = '#8f959e';
exports.udtoken_tag_state_orange_light = '#ff811a';
exports.udtoken_tag_state_purple_light = '#8d55ed';
exports.udtoken_tag_state_red_light = '#e22e28';
exports.udtoken_tag_text_blue = '#002270';
exports.udtoken_tag_text_carmine = '#550c35';
exports.udtoken_tag_text_green = '#022c07';
exports.udtoken_tag_text_indigo = '#151b70';
exports.udtoken_tag_text_lime = '#262e00';
exports.udtoken_tag_text_orange = '#3b1a02';
exports.udtoken_tag_text_purple = '#2f0080';
exports.udtoken_tag_text_red = '#590603';
exports.udtoken_tag_text_s_blue = '#1456f0';
exports.udtoken_tag_text_s_carmine = '#b82879';
exports.udtoken_tag_text_s_green = '#1a7526';
exports.udtoken_tag_text_s_indigo = '#4752e6';
exports.udtoken_tag_text_s_lime = '#5c6d08';
exports.udtoken_tag_text_s_orange = '#c25705';
exports.udtoken_tag_text_s_purple = '#7a35f0';
exports.udtoken_tag_text_s_red = '#c02a26';
exports.udtoken_tag_text_s_turquoise = '#067062';
exports.udtoken_tag_text_s_violet = '#a630a6';
exports.udtoken_tag_text_s_wathet = '#076a94';
exports.udtoken_tag_text_s_yellow = '#ad7a03';
exports.udtoken_tag_text_turquoise = '#02312a';
exports.udtoken_tag_text_violet = '#520052';
exports.udtoken_tag_text_wathet = '#072b3d';
exports.udtoken_tag_text_yellow = '#382201';
exports.udtoken_upload_bg_error = '#eff0f1';
exports.udtoken_upload_mask_img = 'rgba(31, 35, 41, 0.4)';
