// @universe-design/react/lib/skeleton/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Skeleton = void 0;
Object.defineProperty(exports, "SkeletonAvatar", {
  enumerable: true,
  get: function () {
    return _ShapeItem.SkeletonAvatar;
  }
});
Object.defineProperty(exports, "SkeletonContent", {
  enumerable: true,
  get: function () {
    return _ShapeItem.SkeletonContent;
  }
});
Object.defineProperty(exports, "SkeletonParagraph", {
  enumerable: true,
  get: function () {
    return _ShapeItem.SkeletonParagraph;
  }
});
var _clsx = _interopRequireDefault(require("clsx"));
var _react = _interopRequireDefault(require("react"));
var _util = require("../util");
var _configProvider = require("../config-provider");
require("./style");
var _ShapeItem = require("./ShapeItem");
const Skeleton = props => {
  const {
    className,
    loading = true,
    placeholder,
    animation = !_util.__SCREENSHOT__,
    style,
    children
  } = props;
  const cssScope = (0, _configProvider.useCssScope)('skeleton');
  if (loading) {
    return /*#__PURE__*/_react.default.createElement("div", {
      className: (0, _clsx.default)(className, {
        [cssScope()]: true,
        [cssScope('-animation')]: animation
      }),
      style: style
    }, placeholder);
  }
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, children);
};
exports.Skeleton = Skeleton;
