// @universe-design/react/lib/skeleton/ShapeItem.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SkeletonParagraph = exports.SkeletonContent = exports.SkeletonAvatar = void 0;
var _clsx = _interopRequireDefault(require("clsx"));
var _merge = _interopRequireDefault(require("lodash/merge"));
var _react = _interopRequireDefault(require("react"));
var _configProvider = require("../config-provider");
require("./style");
const ShapeItem = props => {
  const {
    className,
    typeName,
    style
  } = props;
  const cssScope = (0, _configProvider.useCssScope)('skeleton__item');
  return /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)([cssScope(), cssScope(`-${typeName}`), className]),
    style: style
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: cssScope('__animation-layer')
  }));
};
const SkeletonContent = props => {
  const {
    className,
    style,
    width,
    height
  } = props;
  // skip undefined
  const mergedStyle = (0, _merge.default)(style, {
    width,
    height
  });
  return /*#__PURE__*/_react.default.createElement(ShapeItem, {
    typeName: "content",
    className: className,
    style: mergedStyle
  });
};
exports.SkeletonContent = SkeletonContent;
const SkeletonAvatar = props => {
  const {
    className,
    style,
    size
  } = props;
  // skip undefined
  const mergedStyle = (0, _merge.default)(style, {
    width: size,
    height: size
  });
  return /*#__PURE__*/_react.default.createElement(ShapeItem, {
    typeName: "avatar",
    className: className,
    style: mergedStyle
  });
};
exports.SkeletonAvatar = SkeletonAvatar;
const SkeletonParagraph = props => {
  const {
    className,
    style,
    rows = 3
  } = props;
  const cssScope = (0, _configProvider.useCssScope)('skeleton');
  return /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)([cssScope('__paragraph'), className]),
    style: style
  }, [...Array(rows)].map((_, index) => /*#__PURE__*/_react.default.createElement(SkeletonContent, {
    key: index
  })));
};
exports.SkeletonParagraph = SkeletonParagraph;
