// @universe-design/react/es/util/assert.js
import { isFunction, isElement } from '@universe-design/react-utilities';
export { isFunction } from '@universe-design/react-utilities';
export function isWindow(node) {
  return node.toString() === '[object Window]';
}
export function getWindow(node) {
  if (!node) {
    return window;
  }
  if (!isWindow(node)) {
    const {
      ownerDocument
    } = node;
    return ownerDocument ? ownerDocument.defaultView || window : window;
  }
  return node;
}
export function isNonNullable(v) {
  if (v !== undefined && v !== null) {
    return true;
  }
  return false;
}

/** @internal is the given object an Object? */
export const isObject = obj => obj !== null && typeof obj === 'object';

/** @internal is the given object/value a promise? */
export const isPromise = value => isObject(value) && isFunction(value.then);
export function isReactInputElement(node) {
  return node.type === 'input';
}
export function isReactHTMLElement(node) {
  return typeof node.type === 'string';
}
export const isString = val => typeof val === 'string';

/**
 * Fork from https://github.com/juggle/resize-observer/blob/c92bcde23f23a5bc7b6d6e32449dec1e540c419f/src/ResizeObservation.ts#L6
 */
export const isSVG = target => target instanceof SVGElement && 'getBBox' in target;
export const isReplacedElement = target => {
  switch (target.tagName) {
    case 'INPUT':
      if (target.type !== 'image') {
        break;
      }
    case 'VIDEO':
    case 'AUDIO':
    case 'EMBED':
    case 'OBJECT':
    case 'CANVAS':
    case 'IFRAME':
    case 'IMG':
      return true;
    default:
  }
  return false;
};
export const isNonReplacedElement = target => isElement(target) && !isSVG(target) && !isReplacedElement(target) && getComputedStyle(target).display === 'inline';
