// @universe-design/react/es/util/dom.js
import { isElement, isHTMLElement } from '@universe-design/react-utilities';
export function getOwnerDocument(node) {
  var _node$ownerDocument;
  return isElement(node) ? (_node$ownerDocument = node.ownerDocument) !== null && _node$ownerDocument !== void 0 ? _node$ownerDocument : document : document;
}
export function isInputElement(el) {
  return isHTMLElement(el) && el.tagName.toLowerCase() === 'input' && 'select' in el;
}
export function contains(parent, child) {
  if (!parent) {
    return false;
  }
  if (parent === child) {
    return true;
  }
  if ('contains' in parent) {
    return parent.contains(child);
  }
  return false;
}
export function eventContains(parent, ev) {
  if (!parent) {
    return false;
  }
  const {
    target
  } = ev;
  if (!isElement(target)) {
    return false;
  }
  if (ev.composedPath) {
    // 处理 shadowRoot 情况
    return ev.composedPath().indexOf(parent) !== -1;
  }
  return contains(parent, target);
}
export function getActiveElement(node) {
  const doc = getOwnerDocument(node);
  return doc === null || doc === void 0 ? void 0 : doc.activeElement;
}
export function getRelatedTarget(event) {
  var _event$target, _event$relatedTarget;
  const target = (_event$target = event.target) !== null && _event$target !== void 0 ? _event$target : event.currentTarget;
  const activeElement = getActiveElement(target);
  return (_event$relatedTarget = event.relatedTarget) !== null && _event$relatedTarget !== void 0 ? _event$relatedTarget : activeElement;
}
export function isHorizontalOverflowed(element) {
  // 这个函数用于判断单行文本是否溢出，只需要判断 x 方向即可
  return element.scrollWidth > element.clientWidth;
}
export const isVerticalOverflowed = ele =>
// 这个函数用于判断多行文本是否溢出，只需要考虑 y 方向
ele.clientHeight < ele.scrollHeight;
export function isOverflowed(element) {
  // 默认的溢出判断逻辑，包含 x 和 y 两个方向，虽然在用户 line-height 设置有问题的时候可能会有一些误判，但是我们认为用户 line-height 小于内容高度是一种「错误」。
  return isHorizontalOverflowed(element) || isVerticalOverflowed(element);
}
