// @universe-design/react/es/util/logger.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
class UniverseUILogger {
  constructor(option) {
    _defineProperty(this, "_scope", void 0);
    this._scope = option.scope;
  }
  formatMessage(message, style) {
    return [`%c [${this._scope}] ${message}`, style];
  }
  info(message) {
    for (var _len = arguments.length, optionalParams = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
      optionalParams[_key - 1] = arguments[_key];
    }
    return console.log(...this.formatMessage(message), ...optionalParams);
  }
  warn(message) {
    for (var _len2 = arguments.length, optionalParams = new Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
      optionalParams[_key2 - 1] = arguments[_key2];
    }
    return console.warn(...this.formatMessage(message), ...optionalParams);
  }
  error(message) {
    for (var _len3 = arguments.length, optionalParams = new Array(_len3 > 1 ? _len3 - 1 : 0), _key3 = 1; _key3 < _len3; _key3++) {
      optionalParams[_key3 - 1] = arguments[_key3];
    }
    return console.error(...this.formatMessage(message), ...optionalParams);
  }
  debug(message) {
    for (var _len4 = arguments.length, optionalParams = new Array(_len4 > 1 ? _len4 - 1 : 0), _key4 = 1; _key4 < _len4; _key4++) {
      optionalParams[_key4 - 1] = arguments[_key4];
    }
    return console.debug(...this.formatMessage(message), ...optionalParams);
  }
}
export const logger = new UniverseUILogger({
  scope: 'UD'
});
