// @universe-design/react/es/util/tabbable.js
import { isHTMLElement } from '@universe-design/react-utilities';
import { getOwnerDocument } from './dom';
export const hasDisplayNone = element => window.getComputedStyle(element).display === 'none';
export const hasTabIndex = element => element.hasAttribute('tabindex');
export const hasNegativeTabIndex = element => hasTabIndex(element) && element.tabIndex === -1;
export function isDisabled(element) {
  return Boolean(element.getAttribute('disabled')) === true || Boolean(element.getAttribute('aria-disabled')) === true;
}
export function isActiveElement(element) {
  const doc = isHTMLElement(element) ? getOwnerDocument(element) : document;
  return doc.activeElement === element;
}
export function hasFocusWithin(element) {
  if (!document.activeElement) {
    return false;
  }
  return element.contains(document.activeElement);
}
export function isHidden(element) {
  if (element.parentElement && isHidden(element.parentElement)) {
    return true;
  }
  return element.hidden;
}
export function isContentEditable(element) {
  const value = element.getAttribute('contenteditable');
  return value !== 'false' && value !== null;
}
export function isFocusable(element) {
  if (!isHTMLElement(element) || isHidden(element) || isDisabled(element)) {
    return false;
  }
  const {
    localName
  } = element;
  if (['input', 'select', 'textarea', 'button'].indexOf(localName) >= 0) {
    return true;
  }
  const others = {
    a: () => element.hasAttribute('href'),
    audio: () => element.hasAttribute('controls'),
    video: () => element.hasAttribute('controls')
  };
  if (localName in others) {
    return others[localName]();
  }
  if (isContentEditable(element)) {
    return true;
  }
  return hasTabIndex(element);
}
export function isTabbable(element) {
  if (!element) {
    return false;
  }
  return isHTMLElement(element) && isFocusable(element) && !hasNegativeTabIndex(element);
}
