// @universe-design/react/lib/shared/components/trigger/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  getAlignFromPlacement: true,
  defaults: true,
  UDTrigger: true
};
exports.defaults = exports.UDTrigger = void 0;
exports.getAlignFromPlacement = getAlignFromPlacement;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _rcTrigger = _interopRequireDefault(require("@universe-design/rc-trigger"));
var _merge = _interopRequireDefault(require("lodash/merge"));
var _cloneDeep = _interopRequireDefault(require("lodash/cloneDeep"));
var _pick = _interopRequireDefault(require("lodash/pick"));
var _reactUtilities = require("@universe-design/react-utilities");
var _hooks = require("../../../config-provider/hooks");
var _util = require("../../../util");
var _hooks2 = require("../../../hooks");
var _useTriggerDisabled = require("./modifiers/useTriggerDisabled");
var _clsx = _interopRequireDefault(require("clsx"));
var _Popup = _interopRequireDefault(require("./Popup"));
var _useTriggerObserver = require("./modifiers/useTriggerObserver");
var _composeRefs = _interopRequireDefault(require("../../../util/composeRefs"));
var _dom = require("../../../util/dom");
var _reactProvider = require("@universe-design/react-provider");
var _types = require("./types");
Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _types[key];
    }
  });
});
/**
 * @docgen
 * @public
 */

function getAlignFromPlacement(builtinPlacements, placementStr, align) {
  const baseAlign = builtinPlacements[placementStr] || {};
  return (0, _merge.default)((0, _cloneDeep.default)(baseAlign), align);
}
const defaults = exports.defaults = {
  mouseEnterDelay: 100,
  mouseLeaveDelay: 100,
  destroyOnHide: {
    keepParent: false
  },
  popupAlign: {}
};
const UDTrigger = exports.UDTrigger = /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
  const rootPrefixCls = (0, _hooks.useClassPrefix)();
  const {
    prefixCls = `${rootPrefixCls}-trigger`,
    disabled = false,
    popupMotion = 'slide-up',
    arrow: arrowProps,
    mouseEnterDelay = defaults.mouseEnterDelay,
    mouseLeaveDelay = defaults.mouseLeaveDelay,
    popupPlacement,
    popupAlign: popupAlignProp = defaults.popupAlign,
    builtinPlacements,
    modifiers = {},
    destroyOnHide = defaults.destroyOnHide,
    visible: visibleProp,
    checkOverflow = false,
    defaultVisible = false,
    onVisibleChange,
    openClassName = `${prefixCls}-open`,
    className,
    udTriggerRef,
    getPopupContainer: getPopupContainerProp,
    popupClassName,
    children,
    __screenshotAlwaysOpenEnabled = false,
    popupStopPropagation,
    ...restProps
  } = props;
  const {
    popupOverflowAuto: enablePopupOverflowAuto = false,
    forcePopupAlign: enableforcePopupAlign = false
  } = modifiers;

  // getPopupContainer
  const ctxGetPopupContainer = (0, _reactProvider.usePopupPortalContainer)();

  // 获取用于对齐的 popupAlign 配置，逻辑从 rc-trigger 提升
  const popupAlign = (0, _react.useMemo)(() => {
    if (popupPlacement && builtinPlacements) {
      return getAlignFromPlacement(builtinPlacements, popupPlacement, popupAlignProp);
    }
    return popupAlignProp;
  }, [builtinPlacements, popupAlignProp, popupPlacement]);

  // visible
  const [visible, setVisible] = (0, _reactUtilities.useControlledState)(defaultVisible, visibleProp, {
    onChange: onVisibleChange
  });
  const triggerRef = (0, _react.useRef)(null);
  const checkVisible = (0, _react.useCallback)(_visible => {
    if (_visible && checkOverflow) {
      var _triggerRef$current;
      const child = (_triggerRef$current = triggerRef.current) === null || _triggerRef$current === void 0 ? void 0 : _triggerRef$current.getRootDomNode();
      const overflowChecker = typeof checkOverflow === 'function' ? checkOverflow : _dom.isOverflowed;
      if (!child || !overflowChecker(child)) {
        setVisible(false);
        return;
      }
    }
    setVisible(_visible);
  }, [checkOverflow, setVisible]);
  _util.__SCREENSHOT__ && __screenshotAlwaysOpenEnabled &&
  // eslint-disable-next-line react-hooks/rules-of-hooks
  _react.default.useEffect(() => {
    var _triggerRef$current2;
    (_triggerRef$current2 = triggerRef.current) === null || _triggerRef$current2 === void 0 ? void 0 : _triggerRef$current2.forcePopupAlign();
    checkVisible(true);
  });

  // destroy
  let destroyPopupOnHide = false;
  let autoDestroy = false;
  if (typeof destroyOnHide === 'boolean') {
    destroyPopupOnHide = destroyOnHide;
  } else if (destroyOnHide && typeof destroyOnHide === 'object') {
    const {
      keepParent
    } = destroyOnHide;
    destroyPopupOnHide = keepParent === true;
    autoDestroy = keepParent === false;
  }

  // 预处理 props
  const rcTriggerProps = {
    ...restProps,
    children: (0, _util.isFunction)(children) ? children({
      visible
    }) : children,
    popupClassName: (0, _clsx.default)(popupClassName, mouseLeaveDelay === 0 && `${prefixCls}-no-entry`),
    // forceRender 的优先级大于 destroyPopupOnHide
    // 俩都传了 true 的时候, 相当于 hide 后不会销毁
    destroyPopupOnHide,
    autoDestroy,
    mouseEnterDelay: mouseEnterDelay / 1000,
    mouseLeaveDelay: mouseLeaveDelay / 1000,
    prefixCls,
    popupMotion: {
      motionName: `${rootPrefixCls}-${popupMotion}`
    },
    popupAlign,
    builtinPlacements,
    popupPlacement,
    popupVisible: visible,
    /**
     * 这里现在有点问题，rc-trigger onPopupVisibleChange 在 visible false 的时候不能触发，所以 checkOverflow 开启时，如果子元素从 overflow 变化到不 overflow，鼠标移出 trigger 区域，子元素再从不 overflow 变成 overflow，tooltip 会再次出现
     */
    onPopupVisibleChange: checkVisible,
    getPopupContainer: getPopupContainerProp !== null && getPopupContainerProp !== void 0 ? getPopupContainerProp : ctxGetPopupContainer
  };

  // 禁用 trigger
  const [newProps] = (0, _useTriggerDisabled.useTriggerDisabled)(rcTriggerProps, {
    disabled
  });
  (0, _react.useImperativeHandle)(udTriggerRef, () => ({
    visible,
    setVisible,
    forcePopupAlign: () => {
      var _triggerRef$current3;
      triggerRef === null || triggerRef === void 0 ? void 0 : (_triggerRef$current3 = triggerRef.current) === null || _triggerRef$current3 === void 0 ? void 0 : _triggerRef$current3.forcePopupAlign();
    }
  }), [visible, setVisible]);
  const [readyObserver, setReadyObserver] = (0, _react.useState)(false);
  const startObserver = (0, _react.useCallback)(() => setReadyObserver(true), []);
  const [childSize, popupSize, viewPointRect] = (0, _useTriggerObserver.useTriggerObserver)({
    triggerRef,
    popupAlign,
    readyObserver,
    enableforcePopupAlign,
    enablePopupOverflowAuto
  });
  (0, _hooks2.useIsomorphicLayoutEffect)(() => {
    if (!visible && readyObserver) {
      setReadyObserver(false);
    }
  }, [visible, readyObserver]);
  const composeTrigger = (0, _react.useMemo)(() => (0, _composeRefs.default)(triggerRef, ref), [triggerRef, ref]);
  if (disabled) {
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, rcTriggerProps.children);
  }

  // TODO 临时逻辑，处理 menu eventKey 透传
  process.env.NODE_ENV !== "production" ? (0, _util.warning)(!('eventKey' in newProps), (0, _util.devWarning)('Trigger',
  // eslint-disable-next-line max-len
  'The `eventKey` property is detected, maybe you wrap the popper elements such as <Tooltip /> or <Popover /> outside the <MenuItem /> or <SubMenu />, it is recommended to move it to the inside')) : void 0;
  newProps.children = /*#__PURE__*/_react.default.cloneElement(newProps.children, (0, _pick.default)(newProps, ['eventKey']));
  return /*#__PURE__*/_react.default.createElement(_rcTrigger.default, (0, _extends2.default)({}, newProps, {
    popup: /*#__PURE__*/_react.default.createElement(_Popup.default, {
      enablePopupOverflowAuto: enablePopupOverflowAuto,
      childSize: childSize,
      popupSize: popupSize,
      viewPointRect: viewPointRect,
      prefixCls: newProps.prefixCls,
      popup: newProps.popup,
      triggerRef: triggerRef,
      popupAlign: popupAlign,
      arrow: arrowProps,
      popupStopPropagation: popupStopPropagation
    }),
    className: (0, _clsx.default)(className, visible && openClassName),
    popupVisible: visible,
    onPopupAlign: startObserver,
    ref: composeTrigger
  }));
});
if (process.env.NODE_ENV !== "production") {
  UDTrigger.displayName = 'UDTrigger';
}
