// @universe-design/react/lib/shared/components/trigger/Popup.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = Popup;
var _react = require("react");
var _util = require("../../../util");
var _function = require("../../../util/function");
var _useTriggerArrow = require("./modifiers/useTriggerArrow");
function Popup(props) {
  const {
    popupAlign,
    arrow,
    prefixCls,
    childSize,
    popupSize,
    viewPointRect,
    enablePopupOverflowAuto = false,
    popupStopPropagation: {
      click: popupClickStopPropagation = false
    } = {}
  } = props;
  let {
    popup: popupElement
  } = props;
  const popupContentStyle = (0, _react.useMemo)(() => {
    var _popupAlign$overflow, _popupAlign$overflow2, _popupAlign$overflow3, _popupAlign$overflow4;
    const style = {};
    if (!enablePopupOverflowAuto) {
      return undefined;
    }

    // 如果两个方向都要翻转就不做 maxWidth 和 maxHeight 的计算了
    if ((_popupAlign$overflow = popupAlign.overflow) !== null && _popupAlign$overflow !== void 0 && _popupAlign$overflow.adjustX && (_popupAlign$overflow2 = popupAlign.overflow) !== null && _popupAlign$overflow2 !== void 0 && _popupAlign$overflow2.adjustY) {
      return undefined;
    }
    if (!viewPointRect) {
      return undefined;
    }
    style.overflow = 'auto';
    if (!((_popupAlign$overflow3 = popupAlign.overflow) !== null && _popupAlign$overflow3 !== void 0 && _popupAlign$overflow3.adjustX)) {
      if (childSize.right < popupSize.left) {
        style.maxWidth = viewPointRect.right - popupSize.left;
      } else if (childSize.left > popupSize.right) {
        style.maxWidth = popupSize.right - viewPointRect.left;
      }
    }
    if (!((_popupAlign$overflow4 = popupAlign.overflow) !== null && _popupAlign$overflow4 !== void 0 && _popupAlign$overflow4.adjustY)) {
      if (childSize.bottom < popupSize.top) {
        style.maxHeight = viewPointRect.bottom - popupSize.top;
      } else if (childSize.top > popupSize.bottom) {
        style.maxHeight = popupSize.bottom - viewPointRect.top;
      }
    }
    return style;
  }, [popupAlign, viewPointRect, popupSize, childSize, enablePopupOverflowAuto]);
  popupElement = (0, _util.isFunction)(popupElement) ? popupElement() : popupElement;
  popupElement = /*#__PURE__*/(0, _react.isValidElement)(popupElement) ? /*#__PURE__*/(0, _react.cloneElement)(popupElement, {
    style: (0, _util.isNonNullable)(popupContentStyle) ? {
      ...popupContentStyle,
      ...popupElement.props.style
    } : {},
    onClick: (0, _function.composeEventHandlers)(e => {
      if (popupClickStopPropagation) {
        e.stopPropagation();
      }
    }, popupElement.props.onClick)
  }) : popupElement;
  popupElement = (0, _useTriggerArrow.useTriggerArrow)({
    childSize,
    popupSize,
    arrow,
    popup: popupElement,
    prefixCls
  });
  return popupElement;
}
