// @universe-design/react/es/cascader/cascade/cascadeUI.js
/* eslint-disable max-lines-per-function */
/* eslint-disable @typescript-eslint/no-non-null-assertion */
import React, { useMemo } from 'react';
import List from '@universe-design/rc-virtual-list';
import { isLeaf } from '../../tree';
import { useCascaderMenuCssScope } from '../common';
import CascaderOption from './Option';
import { HighlightProvider } from '../../shared';
import { useCascadeContext } from './CascadeContext';
export const CascadeUI = () => {
  const cssScope = useCascaderMenuCssScope();
  const {
    doublyTreeNodeList: treeNodeList,
    listHeight: listHeightProp,
    dropdownLevelWidth,
    searchValue,
    searchFlattenTree,
    searching,
    expandedList,
    checkedCache,
    selectedCache,
    levelsListRef,
    fullHeight,
    virtual
  } = useCascadeContext();
  const listHeight = listHeightProp - 2; // 减去 border

  // =========================== List ===========================
  const listRef = React.useRef(null);

  /**
   * ======== Level Dropdown ========
   */

  /**
   * 渲染 level panel
   * @param parentOption level panel 的父级 option
   * @param highlightOption level panel 需要高亮的 option
   */
  const levelRenderer = (parentOption, highlightOption, index) => {
    let options = (parentOption === null || parentOption === void 0 ? void 0 : parentOption.children) || [];
    if (!parentOption) {
      options = treeNodeList.filter(o => !o.parent);
    } else if (isLeaf(parentOption.__node__)) {
      return null;
    }
    return /*#__PURE__*/React.createElement("div", {
      key: (parentOption === null || parentOption === void 0 ? void 0 : parentOption.__node__.value) || 'UD_CASCADER_RESERVED_ROOT_KEY',
      className: cssScope('__level'),
      style: {
        width: dropdownLevelWidth
      }
    }, /*#__PURE__*/React.createElement("div", {
      className: cssScope('__level__list')
    }, /*#__PURE__*/React.createElement(List, {
      itemKey: node => node.__node__.value,
      ref: node => levelsListRef.current[index] = node,
      data: options,
      height: listHeight - 2 * 8 - 2 * 1,
      itemHeight: 32,
      fullHeight: fullHeight
      // onScroll={onPopupScroll}
      ,
      virtual: virtual({
        type: 'levelMenu',
        context: {
          parentOption,
          level: index + 1
        }
      })
      // onMouseEnter={onMouseEnter}
    }, childNode => {
      const {
        __node__: {
          value: key
        }
      } = childNode;
      return (
        /*#__PURE__*/
        // 这里加一个空 div 是为了把内部带 margin 的元素包住，让 virtual list 能正常计算
        React.createElement("div", null, /*#__PURE__*/React.createElement(CascaderOption, {
          isChecked: checkedCache.get(key),
          isSelected: selectedCache.get(key),
          showExpandIcon: true,
          key: key,
          node: childNode,
          highlightOption: highlightOption
        }))
      );
    })));
  };
  let menuNode;
  if (searching) {
    menuNode = /*#__PURE__*/React.createElement("div", {
      className: cssScope('__list')
    }, /*#__PURE__*/React.createElement(List, {
      itemKey: node => node.__node__.value,
      ref: listRef,
      data: searchFlattenTree,
      height: listHeight - 2 * 3 - 2 * 1,
      itemHeight: 32,
      fullHeight: false
      // onScroll={onPopupScroll}
      ,
      virtual: virtual({
        type: 'searchMenu',
        context: undefined
      })
      // onMouseEnter={onMouseEnter}
    }, childNode => {
      const {
        __node__: {
          value: key
        }
      } = childNode;
      return (
        /*#__PURE__*/
        // 这里加一个空 div 是为了把内部带 margin 的元素包住，让 virtual list 能正常计算，tree，select，cascader
        React.createElement("div", null, /*#__PURE__*/React.createElement(CascaderOption, {
          isChecked: checkedCache.get(key),
          isSelected: selectedCache.get(key),
          showExpandIcon: false,
          key: key,
          node: childNode,
          highlightOption: null
        }))
      );
    }));
  } else {
    menuNode = /*#__PURE__*/React.createElement("div", {
      className: cssScope('__level__wrapper')
    }, expandedList.map((_ref, index) => {
      let {
        parent,
        highlight
      } = _ref;
      return levelRenderer(parent, highlight, index);
    }));
  }
  return /*#__PURE__*/React.createElement(HighlightProvider, {
    value: useMemo(() => ({
      searchWords: searchValue ? [searchValue] : undefined,
      disabled: false
    }), [searchValue])
  }, menuNode);
};
