// @universe-design/react/es/cascader/cascade/Option.js
import _extends from "@babel/runtime/helpers/esm/extends";
import * as React from 'react';
/* eslint-disable max-lines-per-function */
/* eslint-disable @typescript-eslint/no-non-null-assertion */
import ListCheckBoldOutlined from '@universe-design/icons-react/ListCheckBoldOutlined';
import RightBoldOutlined from '@universe-design/icons-react/RightBoldOutlined';
import clsx from 'clsx';
import { cssTypographyBody0 } from '../../stylex';
import { Checkbox } from '../../checkbox';
import { CheckedType, isLeaf, isNonLeaf, isCheckedDisabled } from '../../tree';
import { useCascaderMenuCssScope } from '../common';
import { logger } from '../../util';
import composeRefs from '../../util/composeRefs';
import { TextOverflow } from '../../shared/components/text-overflow';
import { useTreeNodeLoadData } from '../../tree/hooks/useTreeNodeLoadData';
import { Loading } from '../../loading';
import { useCascadeContext } from './CascadeContext';
export const CascaderOption = /*#__PURE__*/React.forwardRef((_ref, outerRef) => {
  let {
    isChecked,
    isSelected = false,
    node,
    showExpandIcon,
    highlightOption
  } = _ref;
  const {
    labelRenderer,
    loadedKeys,
    setLoadedKeys,
    onLoad,
    loadData,
    setCurrentExpandKey,
    expandTrigger,
    checkable,
    handleSelectChange,
    handleCheckChange
  } = useCascadeContext();
  const checkboxContainerRef = React.useRef(null);
  const {
    __node__,
    key
  } = node;
  const isHighlight = node === highlightOption && isNonLeaf(node.__node__);
  const {
    value,
    disabled = false
  } = __node__;
  const disableCheckbox = isCheckedDisabled(__node__);
  const cssScope = useCascaderMenuCssScope();
  const {
    loading,
    onInternalExpandedToggle,
    isLeaf: isLoadableLeaf
  } = useTreeNodeLoadData({
    doublyTreeNode: node,
    loadedKeys: loadedKeys || [],
    setLoadedKeys,
    onLoad,
    loadData,
    toggleExpanded: () => setCurrentExpandKey(node.key)
  });
  const waitToLoad = loadData && !(loadedKeys !== null && loadedKeys !== void 0 && loadedKeys.includes(key)) && !isLoadableLeaf;
  /**
   * ======= 自动 scroll 到该元素 ========
   */
  const ref = React.useRef(null);
  React.useEffect(() => {
    const target = ref.current;
    if (isHighlight && target && target.parentElement) {
      const parentTop = target.parentElement.scrollTop;
      const parentBottom = parentTop + target.parentElement.clientHeight;
      const targetTop = target.offsetTop;
      const targetBottom = targetTop + target.clientHeight;
      if (targetTop < parentTop || targetBottom > parentBottom) {
        target.parentElement.scrollTop = target.offsetTop;
      }
    }
  }, [isHighlight]);
  const handleMouseEnter = React.useCallback(() => {
    if (!node || expandTrigger !== 'hover') {
      return;
    }
    if (disabled) {
      return;
    }
    onInternalExpandedToggle();
  }, [disabled, expandTrigger, node, onInternalExpandedToggle]);
  const handleClick = React.useCallback(e => {
    const {
      target
    } = e;
    if ((target instanceof HTMLElement || target instanceof SVGElement) && checkboxContainerRef.current && checkboxContainerRef.current.contains(target)) {
      return;
    }
    if (disabled) {
      return;
    }
    if (waitToLoad) {
      onInternalExpandedToggle();
    } else {
      handleSelectChange(node);
    }
  }, [disabled, waitToLoad, onInternalExpandedToggle, handleSelectChange, node]);
  const handleCheckboxClick = React.useCallback(() => {
    handleCheckChange(value);
  }, [handleCheckChange, value]);
  const rightIconNode = React.useMemo(() => {
    // level 显示的时候非叶子节点显示 expand icon，其他选中态都是显示默认的 checker icon
    if (loading) {
      return /*#__PURE__*/React.createElement(Loading, {
        size: "small"
      });
    }
    if (!isLoadableLeaf && showExpandIcon) {
      return /*#__PURE__*/React.createElement(RightBoldOutlined, {
        className: cssScope('__option__expandIcon')
      });
    }
    if (!checkable) {
      return /*#__PURE__*/React.createElement("span", {
        className: cssScope('__option__checker')
      }, isSelected && /*#__PURE__*/React.createElement(ListCheckBoldOutlined, null));
    }
  }, [checkable, cssScope, isLoadableLeaf, isSelected, loading, showExpandIcon]);

  // ================ Checkbox ===============

  let checkboxProps = {
    checked: false,
    indeterminate: false,
    disabled: disableCheckbox
  };
  if (checkable) {
    if (isChecked === CheckedType.Checked) {
      checkboxProps = {
        checked: true,
        indeterminate: false,
        disabled: disableCheckbox
      };
    } else if (isChecked === CheckedType.Indeterminate) {
      checkboxProps = {
        checked: false,
        indeterminate: true,
        disabled: disableCheckbox
      };
    }
    if (isChecked === undefined) {
      logger.error(`${__node__.value}, isChecked(${isChecked}) is undefined`);
      return null;
    }
  }
  return /*#__PURE__*/React.createElement("div", {
    ref: composeRefs(ref, outerRef),
    onClick: handleClick,
    onMouseEnter: handleMouseEnter // 由于打开下拉菜单时有动画，可能会hover到第一个选项，以至于展开菜单
    ,
    key: value,
    className: clsx(cssScope('__option'), isLeaf(__node__) && cssScope('__option-leaf'), isSelected && cssScope('__option-selected'), isHighlight && cssScope('__option-active'), disabled && cssScope('__option-disabled'))
  }, checkable && /*#__PURE__*/React.createElement(Checkbox, _extends({
    ref: checkboxContainerRef,
    className: cssScope('__option__checkbox')
  }, checkboxProps, {
    onChange: handleCheckboxClick
  })), /*#__PURE__*/React.createElement(TextOverflow, {
    as: "div",
    className: clsx(cssScope('__option__label')),
    lineClamp: 1,
    sx: [cssTypographyBody0]
  }, labelRenderer(node)), rightIconNode);
});
export default CascaderOption;
