// @universe-design/react/es/cascader/cascade/useCascade.js
import * as React from 'react';
import { useControlledState } from '@universe-design/react-utilities';
import { isNonLeaf, isLeaf, CheckedType, useTreeCheck, useTreeSelect, ShowCheckedStrategy } from '../../tree';
import { defaultFilterOption, ifCanBeSelected, defaultVirtual } from '../common';
import { useOptions } from '../hooks/useOptions';
import { checkStrategyShowChild, checkStrategyShowParent } from '../../tree/hooks/checkStrategy';
export const strategyMap = {
  [ShowCheckedStrategy.SHOW_CHILD]: checkStrategyShowChild,
  [ShowCheckedStrategy.SHOW_PARENT]: checkStrategyShowParent
};
function getLastDeepestOption(tree, keys) {
  const filteredNodes = tree.filter(node => keys.includes(node.__node__.value));
  return filteredNodes.length ? filteredNodes.reduce((pre, cur) => {
    if (pre.depth > cur.depth) {
      return pre;
    }
    return cur;
  }, filteredNodes[0]) : undefined;
}
export default function useCascade(props, context, ref) {
  const {
    doublyTreeNodeList,
    doublyTreeNodeMap
  } = context;
  const {
    multiple = false,
    searchValue = '',
    filterOption = defaultFilterOption,
    filterResultSorter,
    filterResultLimit = 50,
    filterResultRenderer,
    expandTrigger = 'click',
    loadedKeys: loadedKeysProp,
    loadData,
    onLoad,
    changeOnSelect = multiple,
    checkable = false,
    listHeight = 256,
    dropdownLevelWidth = 160,
    value: valueProp,
    defaultValue,
    onCheck,
    onSelect,
    showCheckedStrategy = ShowCheckedStrategy.SHOW_CHILD,
    onChange,
    fullHeight = true,
    virtual = defaultVirtual
  } = props;
  const levelsListRef = React.useRef([]);
  const [value, setValue] = useControlledState(defaultValue !== null && defaultValue !== void 0 ? defaultValue : [], valueProp, {
    isControlledOverride: 'value' in props,
    onChange
  });
  const [loadedKeys, setLoadedKeys] = useControlledState([], loadedKeysProp);
  const {
    labelRenderer,
    treeNodeList
  } = useOptions({
    inputValue: searchValue,
    filterOption,
    filterResultSorter,
    filterResultRenderer,
    doublyTreeNodeList
  });

  /**
   * ======== Search Dropdown ========
   */

  const searchFlattenTree = React.useMemo(() => {
    const result = treeNodeList.filter(node => checkable ? true : ifCanBeSelected({
      node: node.__node__,
      selectable: !checkable,
      changeOnSelect,
      multiple
    }));
    if (filterResultLimit === false) {
      return result;
    }
    return result.slice(0, filterResultLimit);
  }, [treeNodeList, filterResultLimit, checkable, changeOnSelect, multiple]);

  /**
   * Check
   */
  const checkedKeys = checkable ? value : [];
  const {
    checkedCache,
    toggleChecked
  } = useTreeCheck({
    doublyTreeNodeMap,
    defaultCheckedKeys: [],
    checkedKeys,
    onCheck: (_checkedKeys, action) => {
      onCheck === null || onCheck === void 0 ? void 0 : onCheck(_checkedKeys, action);
      setValue(_checkedKeys);
    },
    checkStrategy: strategyMap[showCheckedStrategy],
    checkStrictly: false
  });

  /**
   * Select
   */
  const selectedKeys = checkable ? [] : value;
  const {
    selectedCache,
    toggleSelected
  } = useTreeSelect({
    doublyTreeNodeMap,
    defaultSelectedKeys: [],
    selectedKeys,
    onSelect: (_selectedKeys, action) => {
      onSelect === null || onSelect === void 0 ? void 0 : onSelect(_selectedKeys, action);
      setValue(_selectedKeys);
    },
    multiple
  });
  const [currentExpandKey, setCurrentExpandKey] = React.useState(() => {
    let list = [...selectedCache.entries()].filter(_ref => {
      let [_, v] = _ref;
      return v;
    }).map(_ref2 => {
      let [k] = _ref2;
      return k;
    });
    if (checkable) {
      list = [...checkedCache.entries()].filter(_ref3 => {
        let [_, v] = _ref3;
        return v === CheckedType.Checked;
      }).map(_ref4 => {
        let [k] = _ref4;
        return k;
      });
    }
    const highlightOption = getLastDeepestOption(treeNodeList, list);
    return (highlightOption === null || highlightOption === void 0 ? void 0 : highlightOption.__node__.value) || null;
  });
  const handleCheckChange = React.useCallback(key => {
    setCurrentExpandKey(key);
    toggleChecked(key);
  }, [setCurrentExpandKey, toggleChecked]);
  const handleSelectChange = React.useCallback(node => {
    const key = node.__node__.value;
    if (node.__node__.disabled) {
      return;
    }
    // 如果点击的 key 未展开则展开，如果已经展开则收起
    // TODO to be tested
    if (expandTrigger === 'click') {
      setCurrentExpandKey(prevKey => {
        if (prevKey === key) {
          var _node$__parent__$valu, _node$__parent__;
          return (_node$__parent__$valu = (_node$__parent__ = node.__parent__) === null || _node$__parent__ === void 0 ? void 0 : _node$__parent__.value) !== null && _node$__parent__$valu !== void 0 ? _node$__parent__$valu : null;
        }
        return key;
      });
    }
    if (ifCanBeSelected({
      node: node.__node__,
      selectable: !checkable,
      changeOnSelect,
      multiple
    })) {
      toggleSelected(key);
    } else if (checkable && isLeaf(node.__node__)) {
      // 如果不能被点选，并且是叶子节点就勾选
      // TODO to be tested
      toggleChecked(key);
    }
  }, [changeOnSelect, checkable, expandTrigger, multiple, setCurrentExpandKey, toggleChecked, toggleSelected]);

  /**
   * Expand
   */
  const expandedList = React.useMemo(() => {
    const _expandedList = [];
    const currentOption = treeNodeList.find(o => o.__node__.value === currentExpandKey);
    if (currentExpandKey === null || !currentOption) {
      _expandedList.unshift({
        parent: null,
        highlight: null
      });
      return _expandedList;
    }
    if (isNonLeaf(currentOption.__node__)) {
      _expandedList.unshift({
        parent: currentOption,
        highlight: null
      });
    }
    for (let opt = currentOption; opt !== null; opt = opt.parent) {
      _expandedList.unshift({
        parent: opt.parent,
        highlight: opt
      });
    }
    return _expandedList;
  }, [currentExpandKey, treeNodeList]);
  const scrollIntoView = React.useCallback((key, scrollConfig) => {
    const scrollNode = getLastDeepestOption(treeNodeList, [key]) || null;
    if (!scrollNode) {
      return;
    }
    // 将要滚动的节点展开
    setCurrentExpandKey(scrollNode === null || scrollNode === void 0 ? void 0 : scrollNode.key);
    // 展开完成之后，方可进行滚动
    setTimeout(() => {
      for (let opt = scrollNode; opt !== null; opt = opt.parent) {
        var _levelsListRef$curren;
        opt.depth && opt.key && ((_levelsListRef$curren = levelsListRef.current[opt.depth - 1]) === null || _levelsListRef$curren === void 0 ? void 0 : _levelsListRef$curren.scrollTo({
          key: opt.key,
          ...scrollConfig
        }));
      }
    });
  }, [treeNodeList]);
  React.useImperativeHandle(ref, () => ({
    scrollIntoView
  }), [scrollIntoView]);
  return React.useMemo(() => ({
    currentExpandKey,
    setCurrentExpandKey,
    treeNodeList,
    labelRenderer,
    searching: Boolean(searchValue),
    loadedKeys,
    setLoadedKeys,
    expandTrigger,
    loadData,
    onLoad,
    listHeight,
    changeOnSelect,
    checkable,
    dropdownLevelWidth,
    searchFlattenTree,
    multiple,
    value,
    setValue,
    searchValue,
    doublyTreeNodeList,
    doublyTreeNodeMap,
    handleSelectChange,
    handleCheckChange,
    expandedList,
    checkedCache,
    selectedCache,
    levelsListRef,
    scrollIntoView,
    fullHeight,
    virtual
  }), [currentExpandKey, setCurrentExpandKey, treeNodeList, labelRenderer, loadedKeys, setLoadedKeys, expandTrigger, loadData, onLoad, listHeight, changeOnSelect, checkable, dropdownLevelWidth, searchFlattenTree, multiple, value, setValue, searchValue, doublyTreeNodeList, doublyTreeNodeMap, handleSelectChange, handleCheckChange, expandedList, checkedCache, selectedCache, levelsListRef, scrollIntoView, fullHeight, virtual]);
}
