// @universe-design/react/src/message/stories/homeless.stories.tsx
import React, { useRef, useState } from 'react';
import { toast, Button, Toast, Divider } from '../..';
import NoOutlined from '@universe-design/icons-react/NoOutlined';
import { renderGlobalToast } from '..';

export default {
  title: 'Components/Toast/临时示例',
  component: undefined,
};

const overlengthStr = `超过两行将出现...，超过两行将出现，超过两行将出现，
超过两行将出现，超过两行将出现，超过两行将出现，
超过两行将出现，超过两行将出现，超过两行将出现，
超过两行将出现，超过两行将出现，超过两行将出现`;

let timer: any;

const case3Fn = (theme: 'normal' | 'light' = 'normal') => {
  toast.success({
    content: overlengthStr,
    showCloseIcon: true,
    onClose: () => {
      toast.info('onClose');
    },
  });
};

const customerOne = (theme: 'normal' | 'light' = 'normal') => {
  toast.success({
    showCloseIcon: true,
    content: overlengthStr,
    duration: null,
  });
};

export const RenderInComponent = () => {
  const [] = useState(false);
  const ref = useRef<Toast>(null);
  return (
    <div
      className="toast-conatiner"
      style={{
        margin: '50px 50px',
        position: 'relative',
      }}
    >
      <Toast ref={ref} duration={null} />
      <Button
        onClick={() => {
          ref.current?.info('This is in container');
        }}
      >
        触发 Toast
      </Button>
    </div>
  );
};

export const DifferentStates = () => (
  <div>
    {renderGlobalToast()}
    <Button
      onClick={() => {
        toast.config({
          top: 100,
          className: 'custom-classname',
        });
      }}
    >
      更改 Toast 距顶部 100px
    </Button>

    <h2>base - toast.Open</h2>
    <Button
      onClick={() => {
        toast.open({
          key: 'open-custom',
          content: '自定义内容（不会自动关闭）: margin-top: 10vh + 自定义icon',
          style: {
            marginTop: '10vh',
          },
          duration: 0,
          icon: <NoOutlined style={{ verticalAlign: '-0.15em', color: 'red' }} />,
          onClose: () => toast.success('自定义 Message 关闭了！'),
        });
      }}
    >
      点我
    </Button>
    <Button
      onClick={() => {
        toast.destroy('open-custom');
      }}
    >
      关闭
    </Button>
    <h2>type</h2>
    <Divider text="Success" />
    <Button onClick={() => toast.success(':半角字符，:;,.‘“[]「」、|=+-——()*&……%¥#@!~·默认 3 秒关闭')}>
      成功自动关闭
    </Button>
    <Button onClick={() => toast.info(JSON.stringify({ a: 1 }))}>成功自动关闭</Button>
    <Button onClick={() => toast.success('成功，不自动关闭', null)}>成功不自动关闭</Button>
    <Divider text="Info" />
    <Button onClick={() => toast.info('提醒，5s 后自动关闭', 5000)}>提醒</Button>
    <Divider text="Warning" />
    <Button onClick={() => toast.warning('警告，10s 后自动关闭', 10000)}>警告</Button>
    <Divider text="Error" />
    <Button onClick={() => toast.error('错误，15s 后自动关闭', 15000)}>错误</Button>
    <Divider text="Loading" />
    <Button onClick={() => toast.loading('加载')}>加载</Button>

    <h2>超长</h2>
    <Button
      onClick={() =>
        toast.success(
          '成功，文本超过 20 个字符，默认可以手动关闭，并且延长展示时长，每新增 10 个字符长度，增加 150ms，最大 10s',
        )
      }
    >
      自动增加时长
    </Button>
    <Button onClick={() => toast.success(overlengthStr)}>点我</Button>

    <h2>超长-带关闭</h2>
    <Button onClick={() => case3Fn()}>点我</Button>

    <h2>自定义</h2>
    <Divider text="默认不关闭" />
    <Button onClick={() => customerOne()}>点我</Button>
    <Divider text="等待 10 S 关闭" />
    <Button
      onClick={() => {
        toast.info({
          showCloseIcon: true,
          content: '等待 10s 后关闭',
          duration: 10000,
          onClose: () => toast.success('成功了！'),
        });
      }}
    >
      点我
    </Button>

    <Divider text="传入自定义 style" />
    <Button
      onClick={() => {
        toast.error({
          showCloseIcon: true,
          content: '自定义内容（不会自动关闭）: 错误类型 + margin-top: 10vh + 自定义icon',
          style: {
            marginTop: '10vh',
          },
          duration: 0,
          icon: <NoOutlined style={{ verticalAlign: '-0.15em' }} />,
        });
      }}
    >
      点我
    </Button>

    <h2>填充样式</h2>
    <Divider text="Success" />
    <Button
      onClick={() =>
        toast.success({
          content: '成功',
          showCloseIcon: true,
        })
      }
    >
      成功
    </Button>

    <Divider text="Info" />
    <Button onClick={() => toast.info({ content: '提醒', showCloseIcon: true })}>提醒</Button>

    <Divider text="Warning" />
    <Button
      onClick={() =>
        toast.warning({
          content: '警告',
          showCloseIcon: true,
        })
      }
    >
      警告
    </Button>

    <Divider text="Error" />
    <Button onClick={() => toast.error({ content: '错误', showCloseIcon: true })}>错误</Button>

    <Divider text="Loading" />
    <Button onClick={() => toast.loading({ content: '加载' })}>加载</Button>

    <Divider text="超长-填充样式" />
    <Button onClick={() => toast.success({ content: overlengthStr })}>点我</Button>

    <Divider text="超长-填充样式-带关闭" />
    <Button onClick={() => case3Fn('light')}>点我</Button>

    <Divider text="自定义-填充样式-默认不关闭" />
    <Button onClick={() => customerOne('light')}>点我</Button>

    <Divider text="手动更新" />
    <Button
      onClick={() => {
        if (timer) {
          clearTimeout(timer);
        }
        const key = `customer-notice-${new Date().getTime()}`;
        toast.success({ content: overlengthStr, key, duration: null });
        timer = setTimeout(() => {
          toast.error({ content: overlengthStr, key });
        }, 3000);
      }}
    >
      点我
    </Button>
  </div>
);
