// @universe-design/react/src/message/stories/screenshot.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import CAction from '../demo/action';
import CBasic from '../demo/basic';
import CClosable from '../demo/closable';
import CClose from '../demo/close';
import CCloseButton from '../demo/closeButton';
import CDuration from '../demo/duration';
import CHooks from '../demo/hooks';
import CTheme from '../demo/theme';
import CUpdateDuration from '../demo/update-duration';
import CUpdate from '../demo/update';
if (!(CAction as Story).storyName) {
  (CAction as Story).storyName = 'Action';
}
if (!(CBasic as Story).storyName) {
  (CBasic as Story).storyName = 'Basic';
}
if (!(CClosable as Story).storyName) {
  (CClosable as Story).storyName = 'Closable';
}
if (!(CClose as Story).storyName) {
  (CClose as Story).storyName = 'Close';
}
if (!(CCloseButton as Story).storyName) {
  (CCloseButton as Story).storyName = 'Close button';
}
if (!(CDuration as Story).storyName) {
  (CDuration as Story).storyName = 'Duration';
}
if (!(CHooks as Story).storyName) {
  (CHooks as Story).storyName = 'Hooks';
}
if (!(CTheme as Story).storyName) {
  (CTheme as Story).storyName = 'Theme';
}
if (!(CUpdateDuration as Story).storyName) {
  (CUpdateDuration as Story).storyName = 'Update duration';
}
if (!(CUpdate as Story).storyName) {
  (CUpdate as Story).storyName = 'Update';
}

export default {
  title: 'Components/Message/截图测试',
  tags: ['autodocs'],
};

export { CAction, CBasic, CClosable, CClose, CCloseButton, CDuration, CHooks, CTheme, CUpdateDuration, CUpdate };
