// @universe-design/react/es/color-picker/ColorPickerContext.js
import React, { useCallback } from 'react';
import isEqual from 'lodash/isEqual';
import { R500 } from '@universe-design/styles/esm/all.light.atom';
import { createContextHook, useControlledState } from '@universe-design/react-utilities';
import { defaultShowDarkCheckmark, formatColor } from './util';
import { useCssScope } from '../config-provider/hooks';
import { isFunction, isNonNullable } from '../util';
import useSyncState from '../hooks/useSyncState';
import { TinyColor } from '@universe-design/colors';
import { createContext, useContext, useContextSelector } from 'use-context-selector';
import { defaultColor } from './const';
const isColorEqual = (pre, cur) => {
  if (pre === undefined && cur === undefined) {
    return true;
  } else if (pre === undefined || cur === undefined) {
    return false;
  } else {
    return new TinyColor(pre).equals(cur);
  }
};
const {
  useContext: useColorPickerContext,
  useContextSelector: useColorPickerContextSelector,
  Provider: ColorPickerProvider
} = createContextHook('ColorPickerContext', {
  strict: true,
  customizedContextMethods: {
    createContext,
    useContext,
    useContextSelector
  }
});
const defaultTransformDisplayColor = c => c.hex;
function useColorPicker(props) {
  const {
    defaultValue = R500,
    value: valueProp,
    showDarkCheckmark = defaultShowDarkCheckmark,
    prefixCls,
    onChange,
    transformDisplayColor: transformDisplayColorProp = defaultTransformDisplayColor
  } = props;
  const cssScope = useCssScope('color-picker', prefixCls);
  const [controlledValue, _setControlledValue] = useControlledState(defaultValue, valueProp, {
    isEqual,
    isControlledOverride: 'value' in props
  });
  const setControlledValue = useCallback((color, event) => {
    if (isNonNullable(color)) {
      const colorObj = formatColor(color);
      isFunction(onChange) && onChange(colorObj.hex, colorObj, event);
      return;
    }
    isFunction(onChange) && onChange(undefined, undefined, event);
    _setControlledValue(color);
  }, [_setControlledValue, onChange]);
  const [syncValue, setSyncValue] = useSyncState(controlledValue, {
    isEqual: isColorEqual
  });
  const setValue = React.useCallback((_value, event) => {
    setSyncValue(_value);
    setControlledValue(_value, event);
  }, [setControlledValue, setSyncValue]);
  const formattedColor = React.useMemo(() => syncValue !== undefined ? formatColor(syncValue) : syncValue, [syncValue]);
  const transformDisplayColor = React.useCallback(color => formatColor(transformDisplayColorProp(color)), [transformDisplayColorProp]);
  const displayColor = React.useMemo(() => formattedColor && transformDisplayColor(formattedColor), [formattedColor, transformDisplayColor]);
  return {
    value: syncValue,
    setValue,
    showDarkCheckmark,
    rootCssScope: cssScope,
    color: formattedColor,
    hsl: (formattedColor !== null && formattedColor !== void 0 ? formattedColor : defaultColor).hsl,
    hsv: (formattedColor !== null && formattedColor !== void 0 ? formattedColor : defaultColor).hsv,
    rgb: (formattedColor !== null && formattedColor !== void 0 ? formattedColor : defaultColor).rgb,
    hex: (formattedColor !== null && formattedColor !== void 0 ? formattedColor : defaultColor).hex,
    displayColor,
    transformDisplayColor
  };
}
export { useColorPickerContext, ColorPickerProvider, useColorPicker, useColorPickerContextSelector };
