// @universe-design/react/src/carousel/transition-descriptors/fade.ts
import { transitionTimingFunctionStandard } from '../../shared/components/transition/timingFunctionConsts';
import { LinkedList } from '../../util/datastructure/LinkedList';
import type { CarouselManager } from '../CarouselManager';
import type { Disposer, Position, Slide, CarouselTransitionDescriptor } from './utils';
import { isSlideEqual, getSlideFromAbsoluteIndex, getAbsoluteIndexFromSlide } from './utils';

export class FadeCarouselTransitionDescriptor implements CarouselTransitionDescriptor {
  defaults: { duration: number } = {
    duration: 1000,
  };

  private _transitionTimingFunction: string;

  constructor(params: { transitionTimingFunction?: string } = {}) {
    this._transitionTimingFunction = params.transitionTimingFunction ?? transitionTimingFunctionStandard;
  }
  onGotoSlicklyStart(params: { slide: Slide; element: HTMLElement }, manager: CarouselManager): void {
    const { slide, element } = params;
    const {
      state: { slide: curSlide, prevSlide },
    } = manager;
    if ((prevSlide && isSlideEqual(slide, prevSlide)) || (!prevSlide && isSlideEqual(slide, curSlide))) {
      this._show(element);
    } else {
      this._hide(element);
    }
    this._setPosition(element, this._getSlidePosition(slide, prevSlide ?? curSlide, manager));
  }

  onGotoSlicklyActive(params: { slide: Slide; element: HTMLElement }, manager: CarouselManager): void | Disposer {
    const { slide, element } = params;
    const {
      duration,
      state: { slide: curSlide },
    } = manager;
    element.style.willChange = 'opacity';
    element.style.transition = `opacity ${duration}ms ${this._transitionTimingFunction}`;
    if (isSlideEqual(slide, curSlide)) {
      this._show(element);
    } else {
      this._hide(element);
    }
  }

  onGotoSlicklyDone(params: { slide: Slide; element: HTMLElement }, manager: CarouselManager): void {
    const { slide, element } = params;
    const {
      state: { slide: curSlide },
    } = manager;
    element.style.willChange = '';
    element.style.transition = '';
    if (isSlideEqual(slide, curSlide)) {
      this._show(element);
    } else {
      this._hide(element);
    }
  }

  onGotoDirectly(params: { slide: Slide; element: HTMLElement }, manager: CarouselManager): void {
    const { slide, element } = params;
    const {
      state: { slide: curSlide },
    } = manager;
    if (isSlideEqual(slide, curSlide)) {
      this._show(element);
    } else {
      this._hide(element);
    }
    this._setPosition(element, this._getSlidePosition(slide, curSlide, manager));
  }

  getSlidesTOBeRendered(manager: CarouselManager): Slide[] {
    // Constant
    const cacheCount = 1;
    // State
    const {
      slideCount,
      state: { slide: curSlide },
    } = manager;
    const slides = new LinkedList<Slide>();
    for (let i = -cacheCount; i <= cacheCount; i++) {
      const slide = getSlideFromAbsoluteIndex(getAbsoluteIndexFromSlide(curSlide, slideCount) + i, slideCount);
      slides.append(slide);
    }
    return slides.toArray();
  }

  private _getSlidePosition(slide: Slide, curSlide: Slide, manager: CarouselManager): Position {
    const {
      state: { windowSize },
    } = manager;
    return {
      x: windowSize.width / 2 - manager.getSlideSize(slide.slideIndex).width / 2,
      y: 0,
      z: 0,
    };
  }

  private _setPosition(element: HTMLElement, position: Position) {
    const { x, y, z } = position;
    element.style.transform = `translate3d(${x}px, ${y}px, ${z}px)`;
  }

  private _show(element: HTMLElement) {
    element.style.opacity = '1';
    element.style.pointerEvents = '';
  }

  private _hide(element: HTMLElement) {
    element.style.opacity = '0';
    element.style.pointerEvents = 'none';
  }
}
