// @universe-design/react/src/carousel/transition-descriptors/push.ts
import { transitionTimingFunctionStandard } from '../../shared/components/transition/timingFunctionConsts';
import { LinkedList } from '../../util/datastructure/LinkedList';
import type { CarouselManager } from '../CarouselManager';
import type { Position, Slide, CarouselTransitionDescriptor } from './utils';
import { getSlideFromAbsoluteIndex, getAbsoluteIndexFromSlide } from './utils';

export class PushCarouselTransitionDescriptor implements CarouselTransitionDescriptor {
  defaults: { duration: number } = {
    duration: 300,
  };

  private _transitionTimingFunction: string;

  constructor(params: { transitionTimingFunction?: string } = {}) {
    this._transitionTimingFunction = params.transitionTimingFunction ?? transitionTimingFunctionStandard;
  }

  getSlidesTOBeRendered(manager: CarouselManager): Slide[] {
    // Constant
    const cacheWindowTimes = 3;
    // State
    const {
      slideCount,
      state: { slide: curSlide, windowSize },
    } = manager;
    const slides = new LinkedList([curSlide]);
    const curSlideWidth = manager.getSlideSize(curSlide.slideIndex).width;
    // 从中线开始往两边扩散，停止扩散的条件
    // 1. 遇到当前 slide 宽度是 0 说明还没有渲染过，停止扩散
    // 2. 大于等于 cacheWindowTimes / 2 倍的 windowSize 时，停止扩散
    for (const direction of [-1, 1]) {
      for (let _width = curSlideWidth / 2, _slide = curSlide; ; ) {
        if (_width >= (cacheWindowTimes / 2) * windowSize.width) {
          break;
        }
        _slide = getSlideFromAbsoluteIndex(getAbsoluteIndexFromSlide(_slide, slideCount) + direction, slideCount);
        const _slideSize = manager.getSlideSize(_slide.slideIndex);
        if (direction > 0) {
          slides.append(_slide);
        } else {
          slides.prepend(_slide);
        }
        if (_slideSize.width === 0) {
          break;
        }
        _width += _slideSize.width;
      }
    }
    return slides.toArray();
  }

  onGotoSlicklyStart(params: { slide: Slide; element: HTMLElement }, manager: CarouselManager): void {
    const { slide, element } = params;
    const {
      state: { slide: curSlide, prevSlide },
    } = manager;
    this._setPosition(element, this._getSlidePosition(slide, prevSlide ?? curSlide, manager));
  }

  onGotoSlicklyActive(params: { slide: Slide; element: HTMLElement }, manager: CarouselManager): void | (() => void) {
    const { slide, element } = params;
    const {
      duration,
      state: { slide: curSlide },
    } = manager;
    element.style.willChange = 'transform';
    element.style.transition = `transform ${duration}ms ${this._transitionTimingFunction}`;
    this._setPosition(element, this._getSlidePosition(slide, curSlide, manager));
  }

  onGotoSlicklyDone(params: { slide: Slide; element: HTMLElement }, manager: CarouselManager): void {
    const { element } = params;
    element.style.willChange = '';
    element.style.transition = '';
  }

  onGotoDirectly(params: { slide: Slide; element: HTMLElement }, manager: CarouselManager): void {
    const { slide, element } = params;
    const {
      state: { slide: curSlide },
    } = manager;
    this._setPosition(element, this._getSlidePosition(slide, curSlide, manager));
  }

  private _setPosition = (element: HTMLElement, position: Position) => {
    const { x, y, z } = position;
    element.style.transform = `translate3d(${x}px, ${y}px, ${z}px)`;
  };

  private _getSlidePosition = (slide: Slide, curSlide: Slide, manager: CarouselManager): Position => {
    const {
      slideCount,
      state: { windowSize },
    } = manager;
    const slideAbsoluteIndex = getAbsoluteIndexFromSlide(slide, slideCount);
    const curSlideAbsoluteIndex = getAbsoluteIndexFromSlide(curSlide, slideCount);
    // -7 [  4  ] -3 [  6  ] 3 [  2  ]
    // pos(n) - pos(n - 1) = w(n - 1)
    // pos(0) = - w(0) / 2
    const position: Position = {
      x: 0,
      y: 0,
      z: 0,
    };
    if (slideAbsoluteIndex > curSlideAbsoluteIndex) {
      for (let i = curSlideAbsoluteIndex + 1; i <= slideAbsoluteIndex; i++) {
        const { slideIndex } = getSlideFromAbsoluteIndex(i - 1, slideCount);
        position.x += manager.getSlideSize(slideIndex).width;
      }
    } else {
      for (let i = curSlideAbsoluteIndex - 1; i >= slideAbsoluteIndex; i--) {
        const { slideIndex } = getSlideFromAbsoluteIndex(i, slideCount);
        position.x -= manager.getSlideSize(slideIndex).width;
      }
    }

    position.x -= manager.getSlideSize(curSlide.slideIndex).width / 2; // 减去中心 slide 的一半宽度
    position.x += windowSize.width / 2; // 加上一半的视窗宽度
    return position;
  };
}
