// @universe-design/react/src/carousel/transition-descriptors/utils.ts
import type { CarouselManager } from '../CarouselManager';

export interface Position {
  x: number;
  y: number;
  z: number;
}

export interface Slide {
  // slide 所在组的索引值
  groupIndex: number;
  // slide 在用户提供数据中的索引值
  slideIndex: number;
}

export const isSlideEqual = (s1: Slide, s2: Slide) =>
  s1.groupIndex === s2.groupIndex && s1.slideIndex === s2.slideIndex;

export const getAbsoluteIndexFromSlide = (slide: Slide, slideCount: number) =>
  slide.groupIndex * slideCount + slide.slideIndex;

export const getSlideFromAbsoluteIndex = (absoluteIndex: number, total: number): Slide => {
  const groupIndex = Math.floor(absoluteIndex / total);
  return {
    groupIndex,
    slideIndex: absoluteIndex - total * groupIndex,
  };
};

export interface Size {
  width: number;
  height: number;
}

export type Disposer = () => void;

export interface CarouselTransitionDescriptor {
  defaults: {
    /**
     * 默认的动画时长
     */
    duration: number;
  };
  onGotoSlicklyStart: (params: { slide: Slide; element: HTMLElement }, manager: CarouselManager) => void;
  onGotoSlicklyActive: (params: { slide: Slide; element: HTMLElement }, manager: CarouselManager) => Disposer | void;
  onGotoSlicklyDone: (params: { slide: Slide; element: HTMLElement }, manager: CarouselManager) => void;
  onGotoDirectly: (params: { slide: Slide; element: HTMLElement }, manager: CarouselManager) => void;
  // 获取需要渲染的 slides 列表
  getSlidesTOBeRendered: (manager: CarouselManager) => Slide[];
}
