// @universe-design/styles/esm/ccm-ccm-doc/index.light.atom.js
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
export const ccmtoken_doc_block_bg_area = 'rgba(100, 106, 115, 0.05)';
export const ccmtoken_doc_block_bg_hover = 'rgba(20, 86, 240, 0.1)';
export const ccmtoken_doc_block_bg_selected = 'rgba(20, 86, 240, 0.18)';
export const ccmtoken_doc_blockbackground_blue = '#e1eaff';
export const ccmtoken_doc_blockbackground_blue_solid = '#bacefd';
export const ccmtoken_doc_blockbackground_gray = '#dee0e3';
export const ccmtoken_doc_blockbackground_gray_solid = '#bbbfc4';
export const ccmtoken_doc_blockbackground_green = '#d9f5d6';
export const ccmtoken_doc_blockbackground_green_solid = '#b7edb1';
export const ccmtoken_doc_blockbackground_netural = '#f2f3f5';
export const ccmtoken_doc_blockbackground_netural_solid = '#eff0f1';
export const ccmtoken_doc_blockbackground_orange = '#feead2';
export const ccmtoken_doc_blockbackground_orange_solid = '#fed4a4';
export const ccmtoken_doc_blockbackground_purple = '#ece2fe';
export const ccmtoken_doc_blockbackground_purple_solid = '#cdb2fa';
export const ccmtoken_doc_blockbackground_red = '#fde2e2';
export const ccmtoken_doc_blockbackground_red_solid = '#fbbfbc';
export const ccmtoken_doc_blockbackground_yellow = '#ffffcc';
export const ccmtoken_doc_blockbackground_yellow_solid = '#fffca3';
export const ccmtoken_doc_bottom_like_bg_hover = '#ffd13b';
export const ccmtoken_doc_bottom_like_bg_nomal = '#ffc60a';
export const ccmtoken_doc_button_bg_blue = '#e1eaff';
export const ccmtoken_doc_button_bg_blue_solid = '#3370ff';
export const ccmtoken_doc_button_bg_carmine = '#fdddef';
export const ccmtoken_doc_button_bg_carmine_solid = '#f14ba9';
export const ccmtoken_doc_button_bg_green = '#d9f5d6';
export const ccmtoken_doc_button_bg_green_solid = '#62d256';
export const ccmtoken_doc_button_bg_neutral = '#eff0f1';
export const ccmtoken_doc_button_bg_neutral_solid = '#2b2f36';
export const ccmtoken_doc_button_bg_orange = '#feead2';
export const ccmtoken_doc_button_bg_orange_solid = '#ffa53d';
export const ccmtoken_doc_button_bg_purple = '#ece2fe';
export const ccmtoken_doc_button_bg_purple_solid = '#935af6';
export const ccmtoken_doc_button_bg_red = '#fde2e2';
export const ccmtoken_doc_button_bg_red_solid = '#f76964';
export const ccmtoken_doc_button_bg_wathet = '#d9f3fd';
export const ccmtoken_doc_button_bg_wathet_solid = '#14c0ff';
export const ccmtoken_doc_calloutborder_blue = '#82a7fc';
export const ccmtoken_doc_calloutborder_green = '#8ee085';
export const ccmtoken_doc_calloutborder_orange = '#ffba6b';
export const ccmtoken_doc_calloutborder_purple = '#ad82f7';
export const ccmtoken_doc_calloutborder_red = '#f98e8b';
export const ccmtoken_doc_calloutborder_yellow = '#fff67a';
export const ccmtoken_doc_calloutborder_yellow_solid = '#fff258';
export const ccmtoken_doc_highlightcolor_bg_blue_soft = '#f0f4ff';
export const ccmtoken_doc_highlightcolor_bg_blue_solid = '#e1eaff';
export const ccmtoken_doc_highlightcolor_bg_green_soft = '#f0fbef';
export const ccmtoken_doc_highlightcolor_bg_green_solid = '#d9f5d6';
export const ccmtoken_doc_highlightcolor_bg_neutral_light = 'rgba(245, 246, 247, 0.9)';
export const ccmtoken_doc_highlightcolor_bg_orange_soft = '#fff5eb';
export const ccmtoken_doc_highlightcolor_bg_orange_solid = '#feead2';
export const ccmtoken_doc_highlightcolor_bg_purple_soft = '#f6f1fe';
export const ccmtoken_doc_highlightcolor_bg_purple_solid = '#ece2fe';
export const ccmtoken_doc_highlightcolor_bg_red_soft = '#fef1f1';
export const ccmtoken_doc_highlightcolor_bg_red_solid = '#fde2e2';
export const ccmtoken_doc_highlightcolor_bg_sunflower_soft = '#fefff0';
export const ccmtoken_doc_highlightcolor_bg_sunflower_solid = '#ffffcc';
export const ccmtoken_doc_highlightcolor_blue_soft = 'rgba(186, 206, 253, 0.7)';
export const ccmtoken_doc_highlightcolor_blue_solid = 'rgba(78, 131, 253, 0.55)';
export const ccmtoken_doc_highlightcolor_callout_neutral_light = '#f5f6f7';
export const ccmtoken_doc_highlightcolor_green_soft = 'rgba(183, 237, 177, 0.8)';
export const ccmtoken_doc_highlightcolor_green_solid = '#62d256';
export const ccmtoken_doc_highlightcolor_neutral_light = '#f2f3f5';
export const ccmtoken_doc_highlightcolor_neutral_soft = 'rgba(222, 224, 227, 0.8)';
export const ccmtoken_doc_highlightcolor_neutral_solid = '#bbbfc4';
export const ccmtoken_doc_highlightcolor_orange_soft = 'rgba(254, 212, 164, 0.8)';
export const ccmtoken_doc_highlightcolor_orange_solid = '#ffa53d';
export const ccmtoken_doc_highlightcolor_purple_soft = 'rgba(205, 178, 250, 0.7)';
export const ccmtoken_doc_highlightcolor_purple_solid = 'rgba(147, 90, 246, 0.55)';
export const ccmtoken_doc_highlightcolor_red_soft = '#fbbfbc';
export const ccmtoken_doc_highlightcolor_red_solid = '#f76964';
export const ccmtoken_doc_highlightcolor_sunflower_soft = 'rgba(255, 246, 122, 0.8)';
export const ccmtoken_doc_highlightcolor_sunflower_solid = '#ffe928';
export const ccmtoken_doc_picture_toolbar_bg = 'rgba(0, 0, 0, 0.7)';
export const ccmtoken_doc_tableheader_bg_hover = '#dee0e3';
export const ccmtoken_doc_textcolor_bg_red = 'rgba(245, 74, 69, 0.2)';
export const ccmtoken_doc_textcolor_blue = '#245bdb';
export const ccmtoken_doc_textcolor_callout_gray = '#1f2329';
export const ccmtoken_doc_textcolor_gray = '#8f959e';
export const ccmtoken_doc_textcolor_green = '#2ea121';
export const ccmtoken_doc_textcolor_orange = '#de7802';
export const ccmtoken_doc_textcolor_purple = '#6425d0';
export const ccmtoken_doc_textcolor_red = '#d83931';
export const ccmtoken_doc_textcolor_yellow = '#dc9b04';
export const ccmtoken_doc_toolbar_btn_bg_blue_hover = 'rgba(20, 86, 240, 0.2)';
export const ccmtoken_doc_toolbar_btn_bg_blue_pressed = 'rgba(20, 86, 240, 0.3)';
