// @universe-design/react/es/popconfirm/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React, { forwardRef, useCallback, useEffect, useRef } from 'react';
import clsx from 'clsx';
import stylex from '@universe-design/stylex';
import { Tooltip } from '../tooltip';
import { useCssScope, useLocale } from '../config-provider/hooks';
import { Button } from '../button';
import InfoFilled from '@universe-design/icons-react/InfoFilled';
import SucceedFilled from '@universe-design/icons-react/SucceedFilled';
import WarningFilled from '@universe-design/icons-react/WarningFilled';
import ErrorFilled from '@universe-design/icons-react/ErrorFilled';
import { isFunction, KeyCode } from '../util';
import { getRenderPropValue } from '../util/getRenderPropValue';
import { cssBlockBase, cssTypographyHeadline } from '../stylex';
import './style';
const iconMap = {
  success: SucceedFilled,
  warning: WarningFilled,
  info: InfoFilled,
  error: ErrorFilled
};

// eslint-disable-next-line max-lines-per-function
export const Popconfirm = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    title,
    content,
    icon = 'info',
    trigger = 'click',
    okText,
    cancelText,
    okButtonProps = {},
    cancelButtonProps = {},
    onCancel,
    onConfirm,
    disabled = false,
    mouseLeaveDelay = 100,
    ...restProps
  } = props;
  const locale = useLocale();
  const cssScope = useCssScope('popconfirm');
  const overlayCssScope = cssScope('__overlay');
  const udTriggerRef = useRef(null);
  const renderIcon = () => {
    if (!icon) {
      return null;
    }
    if (typeof icon === 'string' && (icon === 'success' || icon === 'warning' || icon === 'info' || icon === 'error')) {
      const Icon = iconMap[icon];
      return /*#__PURE__*/React.createElement("div", {
        className: cssScope(`-icon-${icon}`)
      }, /*#__PURE__*/React.createElement(Icon, null));
    }
    return /*#__PURE__*/React.createElement("div", {
      className: cssScope('-icon-customer')
    }, /*#__PURE__*/React.createElement("span", {
      className: cssScope('-icon-customer-wrapper')
    }, icon));
  };
  const handleOnConfirm = e => {
    isFunction(okButtonProps.onClick) && okButtonProps.onClick(e);
    isFunction(onConfirm) && onConfirm(e);
    if (!e.isDefaultPrevented() && !disabled) {
      var _udTriggerRef$current;
      (_udTriggerRef$current = udTriggerRef.current) === null || _udTriggerRef$current === void 0 ? void 0 : _udTriggerRef$current.setVisible(false);
    }
  };
  const handleOnCancel = e => {
    isFunction(cancelButtonProps.onClick) && cancelButtonProps.onClick(e);
    isFunction(onCancel) && onCancel(e);
    if (!e.isDefaultPrevented() && !disabled) {
      var _udTriggerRef$current2;
      (_udTriggerRef$current2 = udTriggerRef.current) === null || _udTriggerRef$current2 === void 0 ? void 0 : _udTriggerRef$current2.setVisible(false);
    }
  };
  const handleKeyDown = useCallback(e => {
    if (e.keyCode === KeyCode.ESC && !e.defaultPrevented && !disabled) {
      var _udTriggerRef$current3;
      (_udTriggerRef$current3 = udTriggerRef.current) === null || _udTriggerRef$current3 === void 0 ? void 0 : _udTriggerRef$current3.setVisible(false);
    }
  }, [disabled]);
  useEffect(() => {
    window.addEventListener('keydown', handleKeyDown);
    return () => {
      window.removeEventListener('keydown', handleKeyDown);
    };
  }, [handleKeyDown]);
  const renderButton = () => /*#__PURE__*/React.createElement("div", {
    className: `${overlayCssScope}-button-group`
  }, /*#__PURE__*/React.createElement(Button, _extends({
    size: "small"
  }, cancelButtonProps, {
    onClick: handleOnCancel
  }), cancelText || locale.components.popconfirm.cancelText), /*#__PURE__*/React.createElement(Button, _extends({
    style: {
      marginRight: 0
    },
    type: "primary",
    size: "small"
  }, okButtonProps, {
    onClick: handleOnConfirm
  }), okText || locale.components.popconfirm.okText));
  return /*#__PURE__*/React.createElement(Tooltip, _extends({
    ref: ref,
    udTriggerRef: udTriggerRef,
    prefixCls: cssScope(),
    trigger: trigger,
    overlay: (() => {
      if (!title && !content) {
        return null;
      }
      return /*#__PURE__*/React.createElement("div", {
        className: overlayCssScope
      }, renderIcon(), /*#__PURE__*/React.createElement("div", null, title && /*#__PURE__*/React.createElement("div", {
        className: clsx(`${overlayCssScope}-title`, stylex(cssBlockBase, cssTypographyHeadline))
      }, getRenderPropValue(title)), content && /*#__PURE__*/React.createElement("div", {
        className: `${overlayCssScope}-content`
      }, getRenderPropValue(content)), renderButton()));
    })(),
    disabled: disabled,
    mouseLeaveDelay: mouseLeaveDelay,
    arrowHasStroke: true
  }, restProps));
});
