// @byted-nodex/metrics-codec/lib/typings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FieldType = exports.MetricType = exports.CompressStrategy = void 0;
// 压缩策略
var CompressStrategy;
(function (CompressStrategy) {
    CompressStrategy[CompressStrategy["NO_COMPRESSION"] = 0] = "NO_COMPRESSION";
    CompressStrategy[CompressStrategy["ZLIB"] = 1] = "ZLIB";
    CompressStrategy[CompressStrategy["Zstd"] = 2] = "Zstd";
    CompressStrategy[CompressStrategy["GZIP"] = 3] = "GZIP";
    CompressStrategy[CompressStrategy["LZ4"] = 4] = "LZ4";
    CompressStrategy[CompressStrategy["SNAPPY"] = 5] = "SNAPPY";
})(CompressStrategy || (exports.CompressStrategy = CompressStrategy = {}));
// 数值类型
var MetricType;
(function (MetricType) {
    MetricType[MetricType["Unknown"] = 0] = "Unknown";
    MetricType[MetricType["Store"] = 1] = "Store";
    MetricType[MetricType["RateCounter"] = 2] = "RateCounter";
    MetricType[MetricType["Counter"] = 3] = "Counter";
    MetricType[MetricType["Meter"] = 4] = "Meter";
    MetricType[MetricType["Timer"] = 5] = "Timer";
    MetricType[MetricType["Histogram"] = 6] = "Histogram";
    MetricType[MetricType["MultiFields"] = 7] = "MultiFields";
    MetricType[MetricType["DeltaCounterType"] = 8] = "DeltaCounterType";
})(MetricType || (exports.MetricType = MetricType = {}));
// 0 is float64, 1 is int64
var FieldType;
(function (FieldType) {
    FieldType[FieldType["Float64"] = 0] = "Float64";
    FieldType[FieldType["Int64"] = 1] = "Int64";
})(FieldType || (exports.FieldType = FieldType = {}));
