// @byted-nodex/metrics-codec/lib/util.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkAndSliceStr = exports.getBuffersLen = exports.getBufferStr = exports.float64bits = exports.encodeDouble = exports.encodeAsciiStr = exports.encodeUInt32 = exports.encodeInt32 = exports.encodeInt8 = void 0;
const number_float64_base_to_binary_string_1 = __importDefault(require("@stdlib/number-float64-base-to-binary-string"));
const consts_1 = require("./consts");
function encodeInt8(buf, offset, v) {
    buf.writeInt8(v, offset);
    offset += 1;
    return offset;
}
exports.encodeInt8 = encodeInt8;
function encodeInt32(buf, offset, v) {
    buf.writeInt32LE(v, offset);
    offset += 4;
    return offset;
}
exports.encodeInt32 = encodeInt32;
function encodeUInt32(buf, offset, v) {
    buf.writeUInt32LE(v, offset);
    offset += 4;
    return offset;
}
exports.encodeUInt32 = encodeUInt32;
function encodeAsciiStr(buf, offset, s) {
    const len = s.length;
    buf.writeInt32LE(len, offset);
    offset += 4;
    for (let i = 0; i < len; i++) {
        // metric 中写入的都是 ASCII 字符，这样比 buf.write 快不少
        buf[i + offset] = s.charCodeAt(i);
    }
    offset += len;
    return offset;
}
exports.encodeAsciiStr = encodeAsciiStr;
function encodeDouble(buf, offset, v) {
    buf.writeInt32LE(8, offset);
    offset += 4;
    buf.writeBigUInt64LE(v, offset);
    offset += 8;
    return offset;
}
exports.encodeDouble = encodeDouble;
// 对齐 golang math.Float64bits
function float64bits(v) {
    const binaryStr = (0, number_float64_base_to_binary_string_1.default)(v);
    return BigInt(`0b${binaryStr}`);
}
exports.float64bits = float64bits;
function getBufferStr(bufs) {
    const buf = Buffer.concat(bufs);
    const arr = [];
    buf.forEach(v => arr.push(parseInt(String(v), 10)));
    return arr.join(' ');
}
exports.getBufferStr = getBufferStr;
function getBuffersLen(bufs) {
    return bufs.reduce((prev, cur) => {
        return prev + cur.length;
    }, 0);
}
exports.getBuffersLen = getBuffersLen;
function checkAndSliceStr(s) {
    if (s.length > consts_1.MaxInt32) {
        s = s.slice(0, consts_1.MaxInt32);
    }
    return s;
}
exports.checkAndSliceStr = checkAndSliceStr;
