// @universe-design/react/src/date-picker/utils/date.ts
import moment from 'moment';
import type { MODE, RangeValue } from '../interface';

export function isNullEqual<T>(value1: T, value2: T): boolean | undefined {
  if (!value1 && !value2) {
    return true;
  }
  if (!value1 || !value2) {
    return false;
  }
  return undefined;
}

export function isEqual(
  value1: moment.Moment | undefined,
  value2: moment.Moment | undefined,
  mode: MODE,
  showTime = false,
) {
  const equal = isNullEqual<moment.Moment | undefined>(value1, value2);
  if (typeof equal === 'boolean') {
    return equal;
  }
  return value1?.isSame(value2, showTime ? undefined : mode);
}

export function isValidRangeValue(range: RangeValue) {
  const date1 = range[0];
  const date2 = range[1];

  if ((!date1 && !date2) || (date1 && date2 && date1.isSameOrBefore(date2))) {
    return true;
  }
  return false;
}

export function isEqualRange(range1: RangeValue, range2: RangeValue, mode: MODE) {
  const equal = isNullEqual(range1, range2);
  if (typeof equal === 'boolean') {
    return equal;
  }
  return isEqual(range1[0], range2[0], mode) && isEqual(range1[1], range2[1], mode);
}

export function getWeekStartDate(value: moment.Moment, startOfWeek: number) {
  const startMonth = value.clone().startOf('month');
  const weekFirstDay = startMonth.day();
  return startMonth.clone().subtract((weekFirstDay - startOfWeek + 7) % 7, 'day');
}

export function mergeDateAndTime(date: moment.Moment, time: moment.Moment | null) {
  if (!time) {
    return date;
  }
  return date.clone().hours(time.hours()).minutes(time.minutes()).seconds(time.seconds());
}

/**
 * 当a与b均为undefined时 此时moment(a)与moment(b)是否相等依赖于执行的的时间
 */
export function isSame(a: moment.Moment | undefined, b: moment.Moment | undefined) {
  if (a === undefined && b === undefined) {
    return true;
  }
  if (a !== undefined && b !== undefined) {
    return moment(a).isSame(b);
  }
  return false;
}
