// @universe-design/react/src/date-picker/utils/ui.ts
import type moment from 'moment';
import type { GenerateConfig } from '../../shared/components/picker/generate';
import type { SharedTimeProps } from '../../shared/components/picker/interface';
import type { IFormat, MODE, RangeValue } from '../interface';
import { getYearStart, YEAR_COUNT } from '../panels/base/utils';

export function elementsContains(elements: (HTMLElement | undefined | null)[], target: HTMLElement) {
  return elements.some(ele => ele && ele.contains(target));
}

export function isInHoverRange(
  date: moment.Moment,
  mode: MODE,
  hoverDate?: moment.Moment,
  rangeValue?: RangeValue,
  index: 0 | 1 = 0,
) {
  if (!hoverDate || !rangeValue) {
    return false;
  }

  const [startDate = null, endDate = null] = rangeValue;

  if (index === 0 && endDate) {
    return date.isSameOrAfter(hoverDate, mode) && date.isSameOrBefore(endDate, mode);
  }

  if (index === 1 && startDate) {
    return date.isSameOrAfter(startDate, mode) && date.isSameOrBefore(hoverDate, mode);
  }
  return false;
}

export function getTimeProps<DateType>(
  props: { format?: IFormat } & SharedTimeProps<DateType>,
): SharedTimeProps<DateType> {
  const { format, showHour, showMinute, showSecond, use12Hours } = props;

  const showTimeObj: SharedTimeProps<DateType> = { ...props };

  if (format && typeof format === 'string') {
    if (!format.includes('s') && showSecond === undefined) {
      showTimeObj.showSecond = false;
    }
    if (!format.includes('m') && showMinute === undefined) {
      showTimeObj.showMinute = false;
    }
    if (!format.includes('H') && !format.includes('h') && showHour === undefined) {
      showTimeObj.showHour = false;
    }

    if ((format.includes('a') || format.includes('A')) && use12Hours === undefined) {
      showTimeObj.use12Hours = true;
    }
  }

  // if (picker === 'time') {
  //   return showTimeObj;
  // }

  // if (typeof format === 'function') {
  //   // format of showTime should use default when format is custom format function
  //   delete showTimeObj.format;
  // }

  return showTimeObj;
}

export function getInputSize(format: IFormat, generateConfig: GenerateConfig) {
  let textSize;
  const MIN_INPUT_SIZE = 12;
  if (typeof format === 'function') {
    textSize = format(generateConfig.getNow()).length;
  } else {
    textSize = format.length;
  }
  return Math.max(textSize + 2, MIN_INPUT_SIZE);
}

/**
 * 获取单面板的开始时间
 */
export const getStartOfPanel = (_value: moment.Moment, mode: MODE) => {
  const value = _value.clone();
  if (mode === 'year') {
    return getYearStart(value).startOf('year');
  }

  return value.startOf(mode === 'month' ? 'year' : 'month');
};

/**
 * 判断是否在指定面板上
 */
export function isInRangeView(
  /**
   * 要判断的值
   */
  _value: moment.Moment,
  /**
   * 面板值
   */
  _pickerValues: [moment.Moment, moment.Moment],
  /**
   * 面板 mode
   */
  mode: MODE,
) {
  let unit: moment.DurationInputArg2, startOfPanel: moment.Moment, endOfPanel: moment.Moment;
  const value = _value.clone();

  return _pickerValues.some(_pickerValue => {
    const pickerValue = _pickerValue.clone();

    if (mode === 'year') {
      startOfPanel = getYearStart(pickerValue).startOf('year');
      endOfPanel = startOfPanel
        .clone()
        .add(YEAR_COUNT - 1, 'year')
        .endOf('year');
    } else {
      switch (mode) {
        case 'month':
          unit = 'year';
          break;
        default:
          unit = 'month';
      }
      startOfPanel = pickerValue.clone().startOf(unit);
      endOfPanel = pickerValue.clone().endOf(unit);
    }

    return !value.isBefore(startOfPanel) && !value.isAfter(endOfPanel);
  });
}
