// @byted-service/netpool/lib/consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultLogger = exports.defaultDynamicPoolOptions = exports.defaultUDPPoolOptions = exports.defaultTCPPoolOptions = exports.defaultPoolOptions = exports.noop = exports.version = void 0;
// tslint:disable-next-line: no-var-requires
exports.version = require('../package.json').version;
exports.noop = (() => {
    // do nothing
});
exports.defaultPoolOptions = {
    max: 10,
    min: 0,
    createRetryInterval: 100,
    reapInterval: 10e3,
    throwOnFirstCreateError: false,
    concurrencyTolerance: 1,
    acquireTimeout: 3e3,
    createTimeout: 1e3,
    validateTimeout: 1e3,
    useTimeout: null,
    destroyTimeout: 5e3, // 5s
};
exports.defaultTCPPoolOptions = {
    connectTimeout: 1e3,
    socketDataCheckInterval: 100,
    maxSocketDataCheckTime: 3 * 60 * 1000,
    maxWaitingDestroySocketNum: 100,
    maxBufferSize: 10 * 1024 * 1024,
    maxSocketSurvivalTime: 30 * 60 * 1000,
    maxSocketAge: 10000, // 1w
};
exports.defaultUDPPoolOptions = {
    connectTimeout: 1e3,
    maxSocketSurvivalTime: 30 * 60 * 1000,
    maxSocketAge: 100000, // 10w
};
exports.defaultDynamicPoolOptions = {
    checkInterval: 1000,
    maxFreeTime: 3 * 60 * 1000, // 3min
};
exports.defaultLogger = {
    fatal: console.log.bind(console),
    error: console.log.bind(console),
    warn: console.log.bind(console),
    info: console.log.bind(console),
    debug: console.log.bind(console),
    trace: console.log.bind(console),
};
