// @universe-design/react/lib/table/hooks/useFilter/FilterMenu.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = FilterMenu;
var _react = _interopRequireDefault(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _menu = require("../../../menu");
var _button = require("../../../button");
var _checkbox = require("../../../checkbox");
var _radio = require("../../../radio");
var _empty = require("../../../empty");
var _utils = require("../../utils");
var _list = require("../../../shared/components/list");
function hasSubMenu(filters) {
  return filters.some(_ref => {
    let {
      children
    } = _ref;
    return children;
  });
}
function FilterMenu(props) {
  const {
    prefixCls,
    dropdownPrefixCls,
    tableLocale,
    filters,
    filterMultiple,
    openKeys,
    selectedKeys,
    onMenuClick,
    onSelectKeys,
    onOpenChange,
    onReset,
    onConfirm,
    getPopupContainer
  } = props;
  const renderFilterItems = colFilters => {
    if (colFilters.length === 0) {
      return /*#__PURE__*/_react.default.createElement("div", {
        style: {
          margin: '12px 0'
        }
      }, /*#__PURE__*/_react.default.createElement(_empty.Empty, {
        description: tableLocale.filterEmptyText,
        imageStyle: {
          height: 36
        }
      }));
    }
    return colFilters.map((filter, index) => {
      const key = (0, _utils.serializeFilterKey)(filter.value);
      if (filter.children) {
        return /*#__PURE__*/_react.default.createElement(_menu.Menu.SubMenu, {
          key: key || index,
          title: filter.text,
          popupClassName: `${prefixCls}-dropdown-submenu`
        }, renderFilterItems(filter.children));
      }
      const Component = filterMultiple ? _checkbox.Checkbox : _radio.Radio;
      return /*#__PURE__*/_react.default.createElement(_menu.Menu.Item, {
        key: filter.value !== undefined ? key : index
      }, /*#__PURE__*/_react.default.createElement(Component, {
        checked: selectedKeys.includes(key)
      }), /*#__PURE__*/_react.default.createElement("span", {
        className: `${prefixCls}-dropdown-label`
      }, filter.text));
    });
  };
  const triggerSelect = pramas => {
    var _pramas$selectedKeys;
    onSelectKeys({
      selectedKeys: (_pramas$selectedKeys = pramas.selectedKeys) !== null && _pramas$selectedKeys !== void 0 ? _pramas$selectedKeys : []
    });
  };
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_menu.Menu, {
    multiple: filterMultiple
    // prefixCls={`${dropdownPrefixCls}__menu`}
    ,
    className: (0, _clsx.default)({
      [`${dropdownPrefixCls}__menu-without-submenu`]: !hasSubMenu(filters)
    }),
    mode: "normal",
    onClick: onMenuClick,
    onSelect: triggerSelect,
    onDeselect: triggerSelect,
    selectedKeys: selectedKeys,
    getPopupContainer: getPopupContainer,
    openKeys: openKeys,
    onOpenChange: onOpenChange,
    selectable: true,
    showCheckmark: false
  }, renderFilterItems(filters)), /*#__PURE__*/_react.default.createElement(_list.ListDivider, {
    fluid: true
  }), /*#__PURE__*/_react.default.createElement(_list.ListItemLike, {
    className: `${prefixCls}-dropdown-btns`
  }, /*#__PURE__*/_react.default.createElement(_button.Button, {
    type: "link",
    size: "small",
    disabled: selectedKeys.length === 0,
    onClick: onReset
  }, tableLocale.filterReset), /*#__PURE__*/_react.default.createElement(_button.Button, {
    type: "primary",
    size: "small",
    onClick: onConfirm
  }, tableLocale.filterConfirm)));
}
