// @universe-design/react/lib/table/hooks/useFilter/FilterDropdown.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _isEqual = _interopRequireDefault(require("lodash/isEqual"));
var _FilterOutlined = _interopRequireDefault(require("@universe-design/icons-react/FilterOutlined"));
var _dropdown = require("../../../dropdown");
var _reactUtilities = require("@universe-design/react-utilities");
var _util = require("../../../util");
var _FilterMenu = _interopRequireDefault(require("./FilterMenu"));
var _button = require("../../../button");
var _utils = require("../../utils");
var _List = require("../../../shared/components/list/List");
function FilterDropdown(props) {
  var _filterState$filtered;
  const {
    column,
    prefixCls,
    dropdownPrefixCls,
    tableLocale,
    columnKey,
    filterState,
    triggerFilter,
    children,
    filterMultiple,
    getPopupContainer
  } = props;
  const {
    filterDropdownVisible,
    onFilterDropdownVisibleChange
  } = column;
  const [visible, setVisible] = _react.default.useState(false);
  const clickMenuRef = (0, _react.useRef)(false);
  const forceUpdate = (0, _reactUtilities.useForceUpdate)();
  const filtered = Boolean(filterState && (((_filterState$filtered = filterState.filteredKeys) === null || _filterState$filtered === void 0 ? void 0 : _filterState$filtered.length) || filterState.forceFiltered));

  // dropdown visible
  const mergedVisible = typeof filterDropdownVisible === 'boolean' ? filterDropdownVisible : visible;
  const triggerVisible = newVisible => {
    setVisible(newVisible);
    if (onFilterDropdownVisibleChange) {
      onFilterDropdownVisibleChange(newVisible);
    }
  };

  // Select Keys
  const propFilteredKeys = (0, _react.useMemo)(() => (filterState === null || filterState === void 0 ? void 0 : filterState.filteredKeys) || [], [filterState]);
  const filteredKeysSync = (0, _react.useRef)(propFilteredKeys);
  const getFilteredKeysSync = () => filteredKeysSync.current;
  const setFilteredKeysSync = (0, _react.useCallback)(value => {
    filteredKeysSync.current = value;
    forceUpdate();
  }, [filteredKeysSync, forceUpdate]);
  const onSelectKeys = (0, _react.useCallback)(param => {
    setFilteredKeysSync((0, _utils.unSerializeFilterKeys)(param.selectedKeys));
  }, [setFilteredKeysSync]);
  (0, _react.useEffect)(() => {
    onSelectKeys({
      selectedKeys: (0, _utils.serializeFilterKeys)(propFilteredKeys) || []
    });
  }, [propFilteredKeys, onSelectKeys]);

  // Open Keys
  const [openKeys, setOpenKeys] = (0, _react.useState)([]);

  // Open
  const onOpenChange = keys => {
    setOpenKeys(keys);
  };
  const onMenuClick = () => {
    clickMenuRef.current = true;
  };

  // Submit
  const internalTriggerFilter = keys => {
    triggerVisible(false);
    const mergedKeys = keys && keys.length ? keys : null;
    if (mergedKeys === null && (!filterState || !filterState.filteredKeys)) {
      return null;
    }
    if ((0, _isEqual.default)(mergedKeys, filterState === null || filterState === void 0 ? void 0 : filterState.filteredKeys)) {
      return null;
    }
    triggerFilter({
      column,
      key: columnKey,
      filteredKeys: mergedKeys
    });
  };
  const onConfirm = () => {
    internalTriggerFilter(getFilteredKeysSync());
  };
  const onReset = () => {
    setFilteredKeysSync([]);
    internalTriggerFilter([]);
  };
  let dropdownContent, filterIconComp;
  if ((0, _util.isFunction)(column.filterDropdown)) {
    const selectedKeys = getFilteredKeysSync() || [];
    dropdownContent = column.filterDropdown({
      prefixCls: `${dropdownPrefixCls}-custom`,
      setSelectedKeys: keys => onSelectKeys({
        selectedKeys: (0, _utils.serializeFilterKeys)(keys)
      }),
      selectedKeys,
      confirm: onConfirm,
      clearFilters: onReset,
      filters: column.filters,
      visible: mergedVisible
    });
  } else if (column.filterDropdown) {
    dropdownContent = column.filterDropdown;
  } else {
    const selectedKeys = (0, _utils.serializeFilterKeys)(getFilteredKeysSync() || []);
    dropdownContent = /*#__PURE__*/_react.default.createElement(_FilterMenu.default, {
      prefixCls: `${prefixCls}-filter`,
      tableLocale: tableLocale,
      dropdownPrefixCls: dropdownPrefixCls,
      filterMultiple: filterMultiple,
      filters: column.filters || [],
      openKeys: openKeys,
      selectedKeys: selectedKeys,
      onMenuClick: onMenuClick,
      onSelectKeys: onSelectKeys,
      onOpenChange: onOpenChange,
      onConfirm: onConfirm,
      onReset: onReset,
      getPopupContainer: getPopupContainer
    });
  }
  if ((0, _util.isFunction)(column.filterIcon)) {
    filterIconComp = column.filterIcon(filtered);
  } else if (column.filterIcon) {
    filterIconComp = column.filterIcon;
  } else {
    filterIconComp = /*#__PURE__*/_react.default.createElement(_FilterOutlined.default, null);
  }
  const menu = /*#__PURE__*/_react.default.createElement(_List.ListBox, null, dropdownContent);
  const dropDownNode = /*#__PURE__*/_react.default.createElement(_dropdown.Dropdown, {
    overlay: menu,
    trigger: ['click'],
    visible: mergedVisible,
    onVisibleChange: newVisible => {
      if (newVisible && propFilteredKeys !== undefined) {
        setFilteredKeysSync(propFilteredKeys);
      }
      if (!clickMenuRef.current) {
        triggerVisible(newVisible);
      }
      if (!clickMenuRef.current && !newVisible && !column.filterDropdown) {
        onConfirm();
      }
      clickMenuRef.current = false;
    },
    getPopupContainer: getPopupContainer,
    overlayClassName: `${prefixCls}-filter-dropdown`
  }, /*#__PURE__*/_react.default.createElement(_button.IconButton, {
    color: filtered ? 'primary' : undefined,
    selected: filtered || mergedVisible,
    icon: filterIconComp,
    size: "small"
  }));
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, children, /*#__PURE__*/_react.default.createElement("span", {
    className: (0, _clsx.default)(`${prefixCls}-column-filter`, {
      [`${prefixCls}-filter-trigger-container-open`]: mergedVisible
    }),
    onClick: event => {
      event.stopPropagation();
    }
  }, dropDownNode));
}
var _default = exports.default = FilterDropdown;
