// @universe-design/react/lib/table/hooks/useFilter/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
exports.getFilterData = getFilterData;
var _react = _interopRequireWildcard(require("react"));
var _FilterDropdown = _interopRequireDefault(require("./FilterDropdown"));
var _utils = require("../../utils");
var _clsx = _interopRequireDefault(require("clsx"));
function collectFilterStates(columns, init, pos) {
  let filterStates = [];
  (columns || []).forEach((column, index) => {
    const columnPos = (0, _utils.getColumnPos)(index, pos);
    if ('children' in column) {
      filterStates = [...filterStates, ...collectFilterStates(column.children, init, columnPos)];
    } else if (column.filters || 'filterDropdown' in column || 'onFilter' in column) {
      if ('filteredValue' in column) {
        filterStates.push({
          column,
          key: (0, _utils.getColumnKey)(column, columnPos),
          filteredKeys: column.filteredValue || [],
          forceFiltered: column.filtered
        });
      } else {
        filterStates.push({
          column,
          key: (0, _utils.getColumnKey)(column, columnPos),
          filteredKeys: init && column.defaultFilteredValue ? column.defaultFilteredValue : undefined,
          forceFiltered: column.filtered
        });
      }
    }
  });
  return filterStates;
}
function generateFilterInfo(filterStates) {
  const currentFilters = {};
  filterStates.forEach(_ref => {
    let {
      key,
      filteredKeys,
      column
    } = _ref;
    const {
      filters,
      filterDropdown
    } = column;
    if (filterDropdown) {
      currentFilters[key] = filteredKeys || null;
    } else if (Array.isArray(filteredKeys)) {
      const keys = flattenKeys(filters);
      currentFilters[key] = keys.filter(originKey => filteredKeys.includes(originKey));
    } else {
      currentFilters[key] = null;
    }
  });
  return currentFilters;
}
function flattenKeys(filters) {
  let keys = [];
  (filters || []).forEach(_ref2 => {
    let {
      value,
      children
    } = _ref2;
    keys.push(value);
    if (children) {
      keys = [...keys, ...flattenKeys(children)];
    }
  });
  return keys;
}
function getFilterData(data, filterStates) {
  return filterStates.reduce((currentData, filterState) => {
    const {
      column: {
        onFilter,
        filters
      },
      filteredKeys
    } = filterState;
    if (onFilter && filteredKeys && filteredKeys.length) {
      return currentData.filter(record => filteredKeys.some(key => {
        const keys = flattenKeys(filters);
        const keyIndex = keys.findIndex(k => k === key);
        const realKey = keyIndex !== -1 ? keys[keyIndex] : key;
        return onFilter(realKey, record);
      }));
    }
    return currentData;
  }, data);
}
function useFilter(_ref3) {
  let {
    prefix,
    dropdownPrefix,
    tableLocale,
    mergedColumns,
    onFilterChange,
    getPopupContainer
  } = _ref3;
  const [filterStates, setFilterStates] = (0, _react.useState)(collectFilterStates(mergedColumns, true));
  const mergedFilterStates = (0, _react.useMemo)(() => {
    const collectedStates = collectFilterStates(mergedColumns, false);
    if (collectedStates.every(_ref4 => {
      let {
        filteredKeys
      } = _ref4;
      return filteredKeys === undefined;
    })) {
      return filterStates;
    }
    return collectedStates;
  }, [mergedColumns, filterStates]);
  const triggerFilter = (0, _react.useCallback)(filterState => {
    const newFilterStates = mergedFilterStates.filter(_ref5 => {
      let {
        key
      } = _ref5;
      return key !== filterState.key;
    });
    newFilterStates.push(filterState);
    setFilterStates(newFilterStates);
    onFilterChange(generateFilterInfo(newFilterStates), newFilterStates);
  }, [mergedFilterStates, setFilterStates, onFilterChange]);
  const transformColumns = (0, _react.useCallback)(innerColumns => {
    const injectFilter = (columns, states, pos) => columns.map((column, index) => {
      const columnPos = (0, _utils.getColumnPos)(index, pos);
      const {
        filterMultiple = true
      } = column;
      let newColumn = column;
      if (newColumn.filters || newColumn.filterDropdown) {
        const columnKey = (0, _utils.getColumnKey)(newColumn, columnPos);
        const filterState = states.find(_ref6 => {
          let {
            key
          } = _ref6;
          return columnKey === key;
        });
        newColumn = {
          ...newColumn,
          className: (0, _clsx.default)(newColumn.className, {
            [`${prefix}-column-filter-align-right`]: newColumn.filterIconAlign === 'right'
          }),
          title: renderProps => /*#__PURE__*/_react.default.createElement(_FilterDropdown.default, {
            prefixCls: prefix,
            dropdownPrefixCls: dropdownPrefix,
            tableLocale: tableLocale,
            column: newColumn,
            columnKey: columnKey,
            filterState: filterState,
            filterMultiple: filterMultiple,
            triggerFilter: triggerFilter,
            getPopupContainer: getPopupContainer
          }, (0, _utils.renderColumnTitle)(column.title, renderProps)),
          onHeaderCell: col => {
            const cell = column.onHeaderCell && column.onHeaderCell(col) || {};
            cell.className = (0, _clsx.default)(cell.className, `${prefix}-column-has-filters`);
            return cell;
          }
        };
      }
      if ('children' in newColumn) {
        newColumn = {
          ...newColumn,
          children: injectFilter(newColumn.children, states, columnPos)
        };
      }
      return newColumn;
    });
    return injectFilter(innerColumns, mergedFilterStates);
  }, [prefix, dropdownPrefix, tableLocale, getPopupContainer, triggerFilter, mergedFilterStates]);
  const getFilters = (0, _react.useCallback)(() => generateFilterInfo(mergedFilterStates), [mergedFilterStates]);
  return [transformColumns, mergedFilterStates, getFilters];
}
var _default = exports.default = useFilter;
