// @universe-design/styles/ts/core/index.dark.atom.ts
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
export const coretoken_all_watermark_general = 'rgba(117, 117, 117, 0.08)';
export const coretoken_card_float = '#373737';
export const coretoken_groupshare_bg_placeholder = 'linear-gradient(90deg, #4c88ff 0%, #2e65d1 100%)';
export const coretoken_information_bg_bell = '#373737';
export const coretoken_navigation_btn_bg_normal = 'rgba(240, 240, 240, 0.05)';
export const coretoken_navigation_icon_bg_hover = '#505050';
export const coretoken_newtab_color_app_bg = 'rgba(255, 255, 255, 0.1)';
export const coretoken_settings_bg_keyboard_down_normal = '#373737';
export const coretoken_settings_color_tag_bg_down = '#fcd456';
export const coretoken_settings_color_tag_bg_up = '#ffeaa3';
export const coretoken_settings_color_tag_text = '#63470f';
export const coretoken_settings_notice_banner_bg_general = '#373737';
export const coretoken_settings_shadow_keyboard_down_normal = '#0a0a0a';
export const coretoken_video_component_drag_bar_bg_normal = '#0a0a0a';
