// @universe-design/react-utilities/lib-esm/index.js
/**
 * DOM
 */
export { isClient } from './dom/isClient';
export { isElement } from './dom/isElement';
export { isHTMLElement } from './dom/isHTMLElement';
export { isSVGElement } from './dom/isSVGElement';

/**
 * React Hooks
 */
export { supportRef, updateRef, composeRefs, useMergedRefs } from './hooks/composeRefs';
export { useEventCallback } from './hooks/useEventCallback';
export { useIsomorphicLayoutEffect } from './hooks/useIsomorphicLayoutEffect';
export { useLatest } from './hooks/useLatest';
export { useEffectOnce } from './hooks/useEffectOnce';
export { createContextHook } from '@universe-design/react-provider';
export { useProxyRef } from './hooks/useProxyRef';
export { useLatestState } from './hooks/useLatestState';
export { useResizeObserver, createResizeObserver, getResizeObserver } from './hooks/useResizeObserver';
export { resizeObservable } from './dom/resizeObservable';
export { useControlledState, defaultIsEqual } from './hooks/useControlledState';
export { usePrevious } from './hooks/usePrevious';
export { useStyleSheetOnce, StyleSheetOnce } from './hooks/useStyleSheetOnce';
export { useConstant } from './hooks/useConstant';
export { useForceUpdate } from './hooks/useForceUpdate';
export { useWillUnmount } from './hooks/useWillUnmount';
export { useLogger } from './hooks/useLogger';
export { use } from './hooks/use';
export { useCloneElementRef, FindDOMNode } from './hooks/useCloneElementRef';

/**
 * Domain Event
 */

export { DomainEvent, DomainEventTarget } from './domain-events/dom';
export { useDomainEventTargetContext, DomainEventTargetProvider, DomainEventHandler } from './domain-events/react';

/**
 * Others
 */
export { warn, warning, devWarning, compatibleDevWarning, deprecatedWarning, resetWarned } from './warning';
// 为了兼容历史 API，将来移除
export { compatibleDevWarning as antdCompatibleDevWarning } from './warning';
export { sleep } from './others/sleep';
export { KeyboardKey, forwardRef, getSizeAbbr, isFunction, getSize, microDebounce } from './util';

/**
 * interface
 */

/**
 * Components
 */
export { createInOutTransition } from './components/createInOutTransition';
export { useCheckbox } from './components/checkbox';
export { ResizeObserver } from './components/ResizeObserver';

/**
 * Descendant
 */

export { createDescendantContext, createDescendantContextOptional } from './descendant/use-descendant';
