// @universe-design/react/src/anchor/stories/homeless.stories.tsx
import React, { useState } from 'react';
import type { AnchorRef } from '../..';
import { Anchor, Switch, Layout, getSetScrollTop } from '../..';
import './index.less';

export { default as Fixed } from '../demo/fixed';

export default {
  title: 'Components/Anchor/临时示例',
  component: Anchor,
};

const { Link } = Anchor;

const getContainer = () => document.querySelector<HTMLDivElement>('#content')!;

export const DifferentStates = () => {
  const anchorRef = React.useRef<AnchorRef>(null);
  const [affix, setAffix] = useState<boolean>(false);
  const [simple, setSimple] = useState<boolean>(false);

  const containerScrollTopBeforeAnimatedScroll = React.useRef(0);

  const Hash = React.useCallback(
    ({ href }: { href: string }) => (
      <a
        href={href}
        onClick={ev => {
          containerScrollTopBeforeAnimatedScroll.current = getSetScrollTop(getContainer());
          anchorRef.current?.scrollToSection(href);
        }}
      >
        #
      </a>
    ),
    [],
  );

  React.useEffect(() => {
    const cb = () => {
      // 消除浏览器默认是 hash change scroll
      getSetScrollTop(getContainer(), containerScrollTopBeforeAnimatedScroll.current);
      anchorRef.current?.scrollToSection(location.hash || '#API');
    };
    cb();
    window.addEventListener('hashchange', cb);
    return () => {
      window.removeEventListener('hashchange', cb);
    };
  }, []);

  return (
    <Layout>
      <Layout.Sider>
        <div className="sw" style={{ marginBottom: 20 }}>
          <span className="l">是否开启 affix: </span>
          <Switch checked={affix} onChange={setAffix} />
        </div>
        <div className="sw" style={{ marginBottom: 20 }}>
          <span className="l">无轴样式: </span>
          <Switch checked={simple} onChange={setSimple} />
        </div>
        <div
          style={
            affix
              ? {
                  position: 'absolute',
                  top: 180,
                  right: 24,
                  width: 200,
                }
              : {
                  width: 200,
                }
          }
        >
          <Anchor
            ref={anchorRef}
            className="custom-classname"
            affix={affix}
            simple={simple}
            offsetTop={24}
            getContainer={getContainer}
            onClick={() => {
              containerScrollTopBeforeAnimatedScroll.current = getSetScrollTop(getContainer());
            }}
          >
            <Link className="link-custom-classname" href={`#${encodeURI('基础示例')}`} title="基础示例" />
            <Link href="#components-anchor-demo-static" title="Static demo" />
            <Link href="#API" title="API">
              <Link href="#Anchor-Props" title="Anchor Props" />
              <Link href="#Link-Props" title="Link Props" />
            </Link>
            <Link href="#components-anchor-demo-end" title="Static demo" />
          </Anchor>
        </div>
      </Layout.Sider>
      <Layout.Content id="content" style={{ height: '100vh', overflow: 'auto' }}>
        <div style={{ height: '1000vh' }}>
          <div id={encodeURI('基础示例')} style={{ marginTop: 500 }}>
            <span>基础示例</span>
            <Hash href="#基础示例" />
          </div>
          <div id="components-anchor-demo-static" style={{ marginTop: 500 }}>
            <span>components-anchor-demo-static</span>
            <Hash href="#components-anchor-demo-static" />
          </div>
          <div id="API" style={{ marginTop: 500 }}>
            <span>API</span>
            <Hash href="#API" />
          </div>
          <div id="Anchor-Props" style={{ marginTop: 100 }}>
            <span>Anchor-Props</span>
            <Hash href="#Anchor-Props" />
          </div>
          <div id="Link-Props" style={{ marginTop: 20 }}>
            <span>Link-Props</span>
            <Hash href="#Link-Props" />
          </div>
          <div id="components-anchor-demo-end" style={{ marginTop: 200 }}>
            <span>components-anchor-demo-end</span>
            <Hash href="#components-anchor-demo-end" />
          </div>
        </div>
      </Layout.Content>
    </Layout>
  );
};
