// @universe-design/react/src/anchor/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import CAffix from '../demo/affix';
import CBasic from '../demo/basic';
import CBoundary from '../demo/boundary';
import CHistory from '../demo/history';
import CInk from '../demo/ink';
import COnChange from '../demo/on-change';
import CSimple from '../demo/simple';
if (!(CAffix as Story).storyName) {
  (CAffix as Story).storyName = '1. 静态位置';
  (CAffix as Story).parameters = {
    docs: {
      description: {
        story: '设置`affix`为`false`时，锚点会取消浮动效果，处于原始的固定位置。\n\n',
      },
    },
  };
}
if (!(CBasic as Story).storyName) {
  (CBasic as Story).storyName = '0. 基础用法';
  (CBasic as Story).parameters = {
    docs: {
      description: {
        story: '页面滚动，锚点会浮动在页面固定位置。\n\n',
      },
    },
  };
}
if (!(CBoundary as Story).storyName) {
  (CBoundary as Story).storyName = '4. 设置锚点滚动偏移量';
  (CBoundary as Story).parameters = {
    docs: {
      description: {
        story: '通过`bounds`来设置锚点滚动偏移量。\n\n',
      },
    },
  };
}
if (!(CHistory as Story).storyName) {
  (CHistory as Story).storyName = '6. 点击锚点不记录历史';
  (CHistory as Story).parameters = {
    docs: {
      description: {
        story: '通过自定义`onClick`事件，实现点击不记录历史。\n\n',
      },
    },
  };
}
if (!(CInk as Story).storyName) {
  (CInk as Story).storyName = '3. 隐藏锚点激活状态';
  (CInk as Story).parameters = {
    docs: {
      description: {
        story: '设置 `showInk` 为 `false` 时，会隐藏左侧相关的锚点样式\n\n',
      },
    },
  };
}
if (!(COnChange as Story).storyName) {
  (COnChange as Story).storyName = '5. 监听锚点链接改变';
  (COnChange as Story).parameters = {
    docs: {
      description: {
        story: '当锚点激活链接发生改变时会触发`onChange`事件。\n\n',
      },
    },
  };
}
if (!(CSimple as Story).storyName) {
  (CSimple as Story).storyName = '2. 无轴线模式';
  (CSimple as Story).parameters = {
    docs: {
      description: {
        story: '设置`simple`为`true`时，可以使用无左侧轴线的锚点样式。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Anchor/官网示例',
  tags: ['autodocs'],
};

export { CAffix, CBasic, CBoundary, CHistory, CInk, COnChange, CSimple };
