// @universe-design/react/lib/checkbox/Checkbox.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Checkbox = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _reactUtilities = require("@universe-design/react-utilities");
var _hooks = require("../config-provider/hooks");
var _svgs = require("./_assets/svgs");
var _useCheckboxContext = require("./useCheckboxContext");
var _util = require("../util");
var _markAsDisabledableComp = require("../shared/markAsDisabledableComp");
var _stylex2 = require("../stylex");
require("./style");
// feat 精细的布局，所有样式基于 input dome 的真实状态，具有最高的规范性和一致性

const CheckboxComp = /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
  const {
    onChange,
    onCheckChange,
    onMouseEnter,
    onMouseLeave,
    className,
    children,
    value,
    disabled,
    name,
    error,
    ...otherProps
  } = props;
  const inputRef = (0, _react.useRef)(null);
  const prefix = (0, _hooks.useClassPrefix)();
  const cssScope = (0, _hooks.useCssScope)('checkbox');
  const context = (0, _useCheckboxContext.useCheckboxGroupContext)();
  const checkboxProps = {
    ...otherProps,
    className: (0, _clsx.default)((0, _stylex.default)(_stylex2.cssTypographyBody0), className),
    disabled,
    name,
    error,
    classPrefix: `${prefix}__checkbox`,
    onChange: function (checked, e) {
      const eventParams = {
        target: {
          type: 'checkbox',
          checked,
          value
        },
        stopPropagation() {
          e.stopPropagation();
        },
        preventDefault() {
          e.preventDefault();
        },
        nativeEvent: e.nativeEvent,
        reactEvent: e
      };
      if (context !== null && context !== void 0 && context.onChange) {
        context.onChange(eventParams);
      }
      if ((0, _util.isFunction)(onChange)) {
        onChange(eventParams);
      }
      if ((0, _util.isFunction)(onCheckChange)) {
        onCheckChange(checked);
      }
    }
  };
  if (context) {
    var _context$value;
    checkboxProps.disabled = disabled || context.disabled;
    checkboxProps.name = context.name;
    checkboxProps.error = error || context.error;
    checkboxProps.checked = Boolean((_context$value = context.value) === null || _context$value === void 0 ? void 0 : _context$value.includes(value));
  } else if ('checked' in props) {
    checkboxProps.checked = Boolean(props.checked);
  }
  const {
    checked,
    indeterminate,
    inputProps,
    descriptionProps,
    wrapperProps
  } = (0, _reactUtilities.useCheckbox)(checkboxProps, inputRef);
  (0, _react.useImperativeHandle)(ref, () => inputRef.current);

  /**
   * 由于 background 有渐变动画效果，但 checked-svg 与 indeterminate-svg 的切换是瞬间执行的，所以在切换的瞬间会出现只有 background 没有对勾或横杠的闪黑现象
   * 修改方案是: 同时返回 checked-svg 与 indeterminate-svg，而不是让两者做切换。在 checked 状态下将 checked-svg 的透明度置为 1 而 indeterminate-svg 的透明度置为 0；indeterminate 状态下同理
   */
  return /*#__PURE__*/_react.default.createElement("label", (0, _extends2.default)({}, wrapperProps, {
    className: wrapperProps.className,
    onMouseEnter: onMouseEnter,
    onMouseLeave: onMouseLeave
  }), /*#__PURE__*/_react.default.createElement("span", {
    className: cssScope()
  }, /*#__PURE__*/_react.default.createElement("input", (0, _extends2.default)({}, (0, _util.omit)(otherProps, ['error', 'className', 'style', 'defaultChecked', 'indeterminate']), inputProps, {
    ref: inputRef
  })), /*#__PURE__*/_react.default.createElement("span", {
    className: cssScope('__wallpaper')
  }, /*#__PURE__*/_react.default.createElement(_svgs.CheckboxCheckedSvg, {
    className: (0, _clsx.default)(cssScope('__checked-svg'), checked && !indeterminate && cssScope('__checked-svg-shown'))
  }), /*#__PURE__*/_react.default.createElement(_svgs.CheckboxIndeterminateSvg, {
    className: (0, _clsx.default)(cssScope('__indeterminate-svg'), indeterminate && cssScope('__indeterminate-svg-shown'))
  }))), children ? /*#__PURE__*/_react.default.createElement("span", descriptionProps, children) : null);
});
const Checkbox = exports.Checkbox = /* #__PURE__*/(0, _markAsDisabledableComp.markAsDisabledableComp)(CheckboxComp);
Checkbox.__UD_CHECKBOX = true;
if (process.env.NODE_ENV !== "production") {
  Checkbox.displayName = 'Checkbox';
}
