// @universe-design/react/lib/checkbox/Group.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _clsx = _interopRequireDefault(require("clsx"));
var _react = _interopRequireWildcard(require("react"));
var _reactUtilities = require("@universe-design/react-utilities");
var _hooks = require("../config-provider/hooks");
var _context = require("./context");
var _Checkbox = require("./Checkbox");
var _util = require("../util");
require("./style");
const CheckboxGroup = /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
  var _props$value;
  const groupCssScope = (0, _hooks.useCssScope)('checkbox-group');
  const [value, setValue] = (0, _reactUtilities.useControlledState)(props.defaultValue, (_props$value = props.value) !== null && _props$value !== void 0 ? _props$value : [], {
    isControlledOverride: 'value' in props
  });
  const triggerOnChange = data => {
    (0, _util.isFunction)(props.onChange) && props.onChange(data);
    !('value' in props) && setValue(data);
  };
  const isChecked = data => value === null || value === void 0 ? void 0 : value.includes(data);
  return /*#__PURE__*/_react.default.createElement(_context.CheckboxGroupContextProvider, {
    value: {
      onChange: e => {
        const {
          value: changeValue
        } = e.target;
        // 在 group 模式下，如果没有提供 val，不会触发 onChange。
        if (changeValue === undefined) {
          return;
        }
        if (value) {
          const result = [...value];
          const index = value.indexOf(changeValue);
          if (index === -1) {
            result.push(changeValue);
          } else {
            result.splice(index, 1);
          }
          triggerOnChange(result);
        } else {
          triggerOnChange([changeValue]);
        }
      },
      value,
      name: props.name,
      disabled: props.disabled,
      error: props.error
    }
  }, (() => {
    const {
      className,
      options,
      disabled,
      children,
      style,
      error,
      onMouseEnter,
      onMouseLeave
    } = props;
    let renderCheckbox = children;
    if (options && options.length > 0) {
      renderCheckbox = options.map(option => {
        if (typeof option === 'string') {
          return /*#__PURE__*/_react.default.createElement(_Checkbox.Checkbox, {
            key: option,
            disabled: disabled,
            value: option,
            checked: isChecked(option),
            error: error
          }, option);
        }
        return /*#__PURE__*/_react.default.createElement(_Checkbox.Checkbox, {
          key: `checkobx-option-${option.value}`,
          disabled: option.disabled || disabled,
          value: option.value,
          checked: isChecked(option.value),
          style: option.style,
          className: option.className,
          error: option.error || error
        }, option.label);
      });
    }
    const classNames = (0, _clsx.default)(groupCssScope(), className);
    return /*#__PURE__*/_react.default.createElement("div", {
      className: classNames,
      style: style,
      onMouseEnter: onMouseEnter,
      onMouseLeave: onMouseLeave,
      ref: ref
    }, renderCheckbox);
  })());
});
var _default = exports.default = /*#__PURE__*/_react.default.memo(CheckboxGroup);
