// @universe-design/react/lib/notification/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Notification", {
  enumerable: true,
  get: function () {
    return _Notification.default;
  }
});
exports.getGlobalNotification = getGlobalNotification;
exports.notification = void 0;
exports.renderGlobalNotification = renderGlobalNotification;
exports.useNotification = useNotification;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireWildcard(require("react"));
var _reactDom = _interopRequireDefault(require("react-dom"));
var _configProvider = require("../config-provider");
var _Notification = _interopRequireWildcard(require("./Notification"));
require("./style");
let globalNotification;
function getDefaultContainer() {
  const div = document.createElement('div');
  document.body.appendChild(div);
  return div;
}
function singletonRef(elem) {
  if (elem) {
    globalNotification = elem;
  }
}

/**
 * 用于多次重复 renderGlobalToast 和销毁自动生成的 DOM 元素
 */
class NotificationWrapper extends _react.default.Component {
  constructor() {
    super(...arguments);
    (0, _defineProperty2.default)(this, "div", void 0);
    (0, _defineProperty2.default)(this, "isDefaultContainer", true);
    (0, _defineProperty2.default)(this, "getContainer", () => {
      const {
        getContainer
      } = this.props;
      if (!this.div) {
        const c = getContainer === null || getContainer === void 0 ? void 0 : getContainer();
        this.div = c || getDefaultContainer();
        if (c) {
          this.isDefaultContainer = false;
        }
      }
      return this.div;
    });
  }
  componentWillUnmount() {
    globalNotification = undefined;
    if (this.isDefaultContainer && this.div) {
      this.div.remove();
      this.div = undefined;
    }
  }
  render() {
    return /*#__PURE__*/_react.default.createElement(_Notification.default, (0, _extends2.default)({}, this.props, {
      ref: singletonRef,
      getContainer: this.getContainer
    }));
  }
}
function getGlobalNotification() {
  if (globalNotification) {
    return Promise.resolve(globalNotification);
  }
  return new Promise(resolve => {
    const div = getDefaultContainer();
    // 如果没有使用 useToast 或者 renderGlobalToast 初始化 toast，
    // 则默认取第一个 ConfigProvider 提供的配置
    _reactDom.default.render(/*#__PURE__*/_react.default.createElement(_configProvider.CrossRootConfigProvider, null, /*#__PURE__*/_react.default.createElement(NotificationWrapper, {
      ref: () => {
        if (globalNotification) {
          resolve(globalNotification);
        }
      },
      getContainer: () => div
    })), div);
  });
}
const notification = exports.notification = (0, _Notification.createNotificationWrapper)(getGlobalNotification);
/**
 * 用于创建可以获取 ConfigContext 的全局 Notification 单例
 * @example
 * ```tsx
 * <ConfigProvider>
 *   {renderGlobalNotification({ className: 'my-toast' }, document.getElementById('my-container'))}
 *   <App />
 * </ConfigProvider>
 * ```
 *
 * @param props: ToastProps
 * @param container?: HTMLElement
 * @returns React.ReactPortal
 */
function renderGlobalNotification(props) {
  return /*#__PURE__*/_react.default.createElement(NotificationWrapper, props);
}

/**
 * 获取一个局部的
 * @returns
 */
function useNotification(props, container) {
  const ref = (0, _react.useRef)(null);
  const mountDivRef = (0, _react.useRef)();
  const notificationIns = (0, _Notification.createNotificationWrapper)(() => {
    if (ref.current) {
      return Promise.resolve(ref.current);
    }
    return getGlobalNotification();
  });
  const _getDefaultContainer = () => {
    if (mountDivRef.current) {
      return mountDivRef.current;
    }
    mountDivRef.current = getDefaultContainer();
    return mountDivRef.current;
  };
  return [notificationIns, /*#__PURE__*/_react.default.createElement(_Notification.default, (0, _extends2.default)({
    key: "use-notification"
  }, props, {
    ref: ref,
    getContainer: () => container || _getDefaultContainer()
  }))];
}
