// @universe-design/react/lib/notification/notice.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Notice = void 0;
var _react = _interopRequireWildcard(require("react"));
var _SucceedFilled = _interopRequireDefault(require("@universe-design/icons-react/SucceedFilled"));
var _WarningFilled = _interopRequireDefault(require("@universe-design/icons-react/WarningFilled"));
var _ErrorFilled = _interopRequireDefault(require("@universe-design/icons-react/ErrorFilled"));
var _InfoFilled = _interopRequireDefault(require("@universe-design/icons-react/InfoFilled"));
var _CloseOutlined = _interopRequireDefault(require("@universe-design/icons-react/CloseOutlined"));
var _clsx = _interopRequireDefault(require("clsx"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _stylex2 = require("../stylex");
var _hooks = require("../config-provider/hooks");
var _button = require("../button");
const typeIcons = {
  success: _SucceedFilled.default,
  warning: _WarningFilled.default,
  info: _InfoFilled.default,
  error: _ErrorFilled.default
};
const Notice = exports.Notice = /*#__PURE__*/(0, _react.forwardRef)((_ref, ref) => {
  let {
    btn,
    prefixCls,
    className,
    closable = true,
    closeIcon,
    message,
    description,
    style,
    icon,
    type = 'info',
    content,
    onClick,
    onMouseEnter,
    onMouseLeave,
    onClose
  } = _ref;
  const defaultScope = (0, _hooks.useCssScope)('notification');
  const cssScope = prefixCls || defaultScope();
  const callback = (0, _react.useRef)(onClose);
  const cls = (0, _clsx.default)(cssScope, className, {
    [`${cssScope}__notice`]: true,
    [`${cssScope}-closable`]: closable
  });
  (0, _react.useEffect)(() => {
    callback.current = onClose;
  }, [onClose]);
  const onRequestClose = e => {
    onClose && onClose(e);
  };
  const renderIcon = _react.default.useCallback(() => {
    if (icon) {
      return icon;
    } else if (type) {
      const Icon = typeIcons[type];
      return /*#__PURE__*/_react.default.createElement(Icon, {
        className: `${cssScope}__notice-content-icon`
      });
    }
  }, [icon, type, cssScope]);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: cls,
    onClick: onClick,
    onMouseLeave: onMouseLeave,
    onMouseEnter: onMouseEnter,
    ref: ref
  }, content || /*#__PURE__*/_react.default.createElement("div", {
    style: style,
    className: (0, _clsx.default)(`${cssScope}__notice-content`, `${cssScope}__notice-${type}`)
  }, /*#__PURE__*/_react.default.createElement("span", {
    className: `${cssScope}__notice-border`
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: `${cssScope}__notice-icon-${type ? type : 'default'}`
  }, renderIcon()), /*#__PURE__*/_react.default.createElement("div", {
    className: `${cssScope}__notice-with-icon`
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(`${cssScope}__notice-title`, (0, _stylex.default)(_stylex2.cssTypographyTitle4))
  }, message), /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(`${cssScope}__notice-des`, (0, _stylex.default)(_stylex2.cssTypographyBody0))
  }, description)), btn && /*#__PURE__*/_react.default.createElement("div", {
    className: `${cssScope}__notice-operation`
  }, btn), (() => {
    if (closable) {
      return /*#__PURE__*/_react.default.createElement("span", {
        className: `${cssScope}__notice-close`,
        tabIndex: 0,
        onClick: onRequestClose
      }, closeIcon || /*#__PURE__*/_react.default.createElement(_button.Button, {
        className: `${cssScope}__notice-close-icon`,
        icon: /*#__PURE__*/_react.default.createElement(_CloseOutlined.default, null),
        type: "icon",
        size: "small"
      }));
    }
    return null;
  })()));
});
