// @universe-design/react/lib/notification/Notification.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createNotificationWrapper = createNotificationWrapper;
exports.default = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _pick = _interopRequireDefault(require("lodash/pick"));
var _omit = _interopRequireDefault(require("lodash/omit"));
var _notice = require("./notice");
var _manager = _interopRequireDefault(require("./manager"));
/**
 * 所有命令式调用方法的第一个参数
 */

const ManagerProps = ['className', 'prefixCls', 'duration', 'transitionProps', 'maxCount', 'getContainer', 'top', 'bottom', 'placement'];

/**
 * @docgen
 */

/**
 * @docgen
 */

function createNotificationWrapper(getInstance) {
  return {
    /** 打开通知 */
    open(config) {
      return getInstance().then(t => t.open(config));
    },
    /** 信息通知 */
    info(configOrContent, duration, onClose) {
      return getInstance().then(t => t.info(configOrContent, duration, onClose));
    },
    /** 成功通知 */
    success(configOrContent, duration, onClose) {
      return getInstance().then(t => t.success(configOrContent, duration, onClose));
    },
    /** 警告通知 */
    warning(configOrContent, duration, onClose) {
      return getInstance().then(t => t.warning(configOrContent, duration, onClose));
    },
    /** 同 `warning` */
    warn(configOrContent, duration, onClose) {
      return getInstance().then(t => t.warn(configOrContent, duration, onClose));
    },
    /** 错误通知 */
    error(configOrContent, duration, onClose) {
      return getInstance().then(t => t.error(configOrContent, duration, onClose));
    },
    /** 通知全局配置 */
    config(options) {
      getInstance().then(t => t.config(options));
    },
    /** 同 `destroy` */
    close(key) {
      getInstance().then(t => t.close(key));
    },
    /** 按 `key` 销毁通知，或销毁所有通知 */
    destroy(key) {
      getInstance().then(t => t.destroy(key));
    }
  };
}
function getManagerConfig(option) {
  const {
    key,
    duration,
    onClose,
    placement,
    ...rest
  } = option;
  return {
    key,
    // 如果自定义了操作按钮，则不自动关闭
    duration: rest.btn ? null : duration,
    placement,
    renderMessage: _ref => {
      let {
        close
      } = _ref;
      return /*#__PURE__*/_react.default.createElement(_notice.Notice, (0, _extends2.default)({}, rest, {
        onClose: close
      }));
    },
    onClose
  };
}
class Notification extends _react.default.PureComponent {
  constructor() {
    super(...arguments);
    (0, _defineProperty2.default)(this, "managerRef", /*#__PURE__*/(0, _react.createRef)());
  }
  /**
   * 计算后的配置
   */
  get _config() {
    return Object.assign({}, Notification.globalConfig, this.props);
  }

  /* eslint-disable no-dupe-class-members */
  open(config) {
    const manager = this.managerRef.current;
    const handleClose = config.onClose;
    return new Promise(resolve => {
      config.onClose = () => {
        handleClose && handleClose();
        resolve(undefined);
      };
      (manager === null || manager === void 0 ? void 0 : manager.add(this._getManagerConfig(config))) || '';
    });
  }
  info(configOrContent, duration, onClose) {
    return this._open('info', configOrContent, duration, onClose);
  }
  success(configOrContent, duration, onClose) {
    return this._open('success', configOrContent, duration, onClose);
  }
  warning(configOrContent, duration, onClose) {
    return this._open('warning', configOrContent, duration, onClose);
  }
  warn(configOrContent, duration, onClose) {
    return this.warning(configOrContent, duration, onClose);
  }
  error(configOrContent, duration, onClose) {
    return this._open('error', configOrContent, duration, onClose);
  }
  config(props) {
    Notification.globalConfig = Object.assign(Notification.globalConfig, props);
    // Notification.config = props;
    this.forceUpdate();
  }
  destroy(key) {
    if (key) {
      var _this$managerRef$curr;
      (_this$managerRef$curr = this.managerRef.current) === null || _this$managerRef$curr === void 0 ? void 0 : _this$managerRef$curr.remove(key);
    } else {
      var _this$managerRef$curr2;
      (_this$managerRef$curr2 = this.managerRef.current) === null || _this$managerRef$curr2 === void 0 ? void 0 : _this$managerRef$curr2.destroy();
    }
  }
  close(key) {
    return this.destroy(key);
  }
  render() {
    return /*#__PURE__*/_react.default.createElement(_manager.default, (0, _extends2.default)({
      ref: this.managerRef,
      transitionProps: {
        timeout: 400,
        classNames: 'notification-animation'
      },
      duration: 4000
    }, (0, _pick.default)(this._config, ManagerProps)));
  }
  _getManagerConfig(config) {
    const props = Object.assign((0, _omit.default)(this.props, ManagerProps), (0, _omit.default)(this._config, ManagerProps));
    return getManagerConfig({
      ...props,
      ...config
    });
  }
  _open(type, configOrContent, duration, onClose) {
    if (configOrContent && typeof configOrContent === 'object' && 'message' in configOrContent) {
      return this.open({
        ...configOrContent,
        type
      });
    } else {
      const d = typeof duration === 'function' ? duration() : duration;
      return this.open({
        type,
        message: configOrContent,
        duration: d,
        onClose
      });
    }
  }
  /* eslint-enable no-dupe-class-members */
}
exports.default = Notification;
/**
 * 全局 config
 */
(0, _defineProperty2.default)(Notification, "globalConfig", {});
