// @universe-design/react-ai/lib-commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  AIOnboardingAnimation_unstable: true,
  AIThinkingAnimation: true,
  AIAvatarThinkingAnimation: true,
  AIDotLoading: true,
  AISimplifyThinkingAnimation: true,
  StarThinkingAnimation: true
};
Object.defineProperty(exports, "AIAvatarThinkingAnimation", {
  enumerable: true,
  get: function () {
    return _avatarThinkingAnimation.AIAvatarThinkingAnimation;
  }
});
Object.defineProperty(exports, "AIDotLoading", {
  enumerable: true,
  get: function () {
    return _dotLoading.AIDotLoading;
  }
});
Object.defineProperty(exports, "AIOnboardingAnimation_unstable", {
  enumerable: true,
  get: function () {
    return _onboardingAnimation.AIOnboardingAnimation_unstable;
  }
});
Object.defineProperty(exports, "AISimplifyThinkingAnimation", {
  enumerable: true,
  get: function () {
    return _simplifyThinkingAnimation.AISimplifyThinkingAnimation;
  }
});
Object.defineProperty(exports, "AIThinkingAnimation", {
  enumerable: true,
  get: function () {
    return _thinkingAnimation.AIThinkingAnimation;
  }
});
Object.defineProperty(exports, "StarThinkingAnimation", {
  enumerable: true,
  get: function () {
    return _starThinkingAnimation.StarThinkingAnimation;
  }
});
var _onboardingAnimation = require("./onboarding-animation");
var _thinkingAnimation = require("./thinking-animation");
var _avatarThinkingAnimation = require("./avatar-thinking-animation");
var _dotLoading = require("./dot-loading");
var _simplifyThinkingAnimation = require("./simplify-thinking-animation");
var _button = require("./button");
Object.keys(_button).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _button[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _button[key];
    }
  });
});
var _smoothOutput = require("./smooth-output");
Object.keys(_smoothOutput).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _smoothOutput[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _smoothOutput[key];
    }
  });
});
var _card = require("./card");
Object.keys(_card).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _card[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _card[key];
    }
  });
});
var _starThinkingAnimation = require("./star-thinking-animation");
