// @universe-design/react/es/shared/components/picker/Comp/PickerTrigger.js
import React from 'react';
import clsx from 'clsx';
import { isFunction } from '../../../../util';
import { useCssScope, useDirection } from '../../../../config-provider/hooks';
import { usePopperSidecar } from '../../../../config-provider/hooks/usePopperSidecar';
import { SlideUpDownTransition } from '../../transition/SlideUpDownTransition';
import { usePopupPortalContainer } from '@universe-design/react-provider';
const PickerTrigger = function (props) {
  const {
    disabled,
    open,
    children,
    dropdownClassName,
    popupStyle,
    popupElement,
    setOpened,
    getPopupContainer
  } = props;
  const triggerElementEnable = Boolean(isFunction(children) || children) && !disabled;
  const cssScope = useCssScope('picker');
  const prefixCls = cssScope();
  const Popper = usePopperSidecar();
  const showAction = triggerElementEnable ? ['click'] : [];
  const direction = useDirection();
  const ctxGetPopupContainer = usePopupPortalContainer();
  return /*#__PURE__*/React.createElement(Popper, {
    action: undefined,
    showAction: showAction,
    hideAction: [],
    destroyOnHide: true,
    popupStyle: popupStyle,
    getPopupContainer: triggerNode => {
      const _getPopupContainer = getPopupContainer !== null && getPopupContainer !== void 0 ? getPopupContainer : ctxGetPopupContainer;
      return _getPopupContainer(triggerNode);
    },
    prefixCls: prefixCls,
    popupPlacement: direction === 'rtl' ? 'bottomRight' : 'bottomLeft',
    visible: open,
    popupClassName: clsx(dropdownClassName, `${prefixCls}-dropdown`),
    onVisibleChange: setOpened,
    popup: popupElement,
    popupMotion: "slide-up",
    enableRestoreFocus: false,
    initialFocus: null,
    showArrow: false,
    arrowHasStroke: false,
    Transition: SlideUpDownTransition,
    enableMaxHeight: false,
    defaultMainAxisOffsetWithoutArrow: 4
  }, children);
};
export default PickerTrigger;
