// @universe-design/react/src/transfer/utils/index.ts
import escapeRegExp from 'lodash/escapeRegExp';
import type { TransferGroupOption, TransferOption, TransferValue } from '../interface';

export enum TransferListType {
  CAPION,
  OPTION,
}
/**
 * List 只接受一维数组，这里是拍平后的数据结构
 */
export type FlatItem<V extends TransferValue = string, OP extends TransferOption<V> = TransferOption<V>> =
  | { type: TransferListType.OPTION; option: OP }
  | {
      type: TransferListType.CAPION;
      group: Omit<TransferGroupOption<V, OP>, 'children'>;
    };

/**
 * 将带分组的嵌套数据展平，
 * 因为 List 组件只接受一维数组
 */
export const flatTransferGroups = <V extends TransferValue, OP extends TransferOption<V>>(
  options: OP[],
  groupWeakMap: WeakMap<OP, TransferGroupOption<V, OP>>,
  showGroup: boolean,
): FlatItem<V, OP>[] => {
  if (!showGroup) {
    return options.map(op => ({
      option: op,
      type: TransferListType.OPTION,
    }));
  }

  const grouped: Map<TransferGroupOption<V, OP>, OP[]> = new Map();
  const groupSet = new Set<TransferGroupOption<V, OP>>();

  options.forEach(op => {
    const groupEntity = groupWeakMap.get(op);
    if (!groupEntity) {
      return;
    } else {
      groupSet.add(groupEntity);
      if (!grouped.has(groupEntity)) {
        grouped.set(groupEntity, []);
      }
      grouped.set(groupEntity, [...(grouped.get(groupEntity) || []), op]);
    }
  });
  const data: FlatItem<V, OP>[] = [];

  groupSet.forEach(groupEntity => {
    const groupTitle = groupEntity.title;

    data.push({
      group: {
        title: groupTitle,
        key: groupEntity.key,
      },
      type: TransferListType.CAPION,
    });

    data.push(
      ...(grouped.get(groupEntity) || []).map(op => ({
        option: op,
        type: TransferListType.OPTION as const,
      })),
    );
  });
  return data;
};

export const defaultFilterOption = <V extends TransferValue, OP extends TransferOption<V>>(
  searchValue: string,
  option: OP,
): boolean => {
  const keyword = typeof option.label === 'string' ? option.label : String(option.value);
  return !searchValue || new RegExp(escapeRegExp(searchValue), 'gi').test(keyword);
};

export const optionLabel = <V extends TransferValue>(option: TransferOption<V>) =>
  typeof option.label === 'string' ? option.label : String(option.value);
