// @universe-design/react/src/upload/uploadFile/CardFile.tsx
import clsx from 'clsx';
import React from 'react';
import stylex from '@universe-design/stylex';
import { UploadProgress } from './UploadProgress';
import { isImage } from '../utils';
import { FileContextProvider } from './FileContext';
import { useFile } from './useFile';
import type { FileProps, UploadFileStatus } from './interface';
import { FileName } from './FileName';
import '../style/card-file.less';
import { devWarning, warning } from '../../util';

export type CardFileProps = FileProps;

export const CardFile = (fileProps: CardFileProps) => {
  const { style, nameRenderer } = fileProps;
  const fileCtxValue = useFile(fileProps);
  const {
    classNames,
    file,
    errorNode,
    retryNode,
    removeNode,
    downloadNode,
    previewNode,
    iconRender,
    rootPrefixCls,
    progressProps,
  } = fileCtxValue;

  warning(
    !(['downloading', 'downloadFailed', 'downloaded'] as UploadFileStatus[]).includes(file.status),
    devWarning('Upload', `file.status=${file.status} does not support for CardFile.`),
  );

  function preview() {
    const previewFile = file;
    const previewUrl = previewFile.thumbUrl || previewFile.url;
    if (isImage(previewFile) && previewUrl && previewFile.status === 'success') {
      return (
        <div className={`${rootPrefixCls}-preview-img`}>
          <img src={previewUrl} />
        </div>
      );
    }
    return <div className={`${rootPrefixCls}-preview-icon`}>{iconRender(previewFile)}</div>;
  }

  return (
    <FileContextProvider value={fileCtxValue}>
      <div className={clsx(classNames, `${rootPrefixCls}--card`)} style={style}>
        <div className={clsx(`${rootPrefixCls}__preview`, stylex(stylex.create({ position: 'relative' })))}>
          {preview()}
          <UploadProgress
            file={file}
            progressProps={progressProps}
            prefixCls={rootPrefixCls}
            className={stylex(
              stylex.create({
                position: 'absolute',
                bottom: 0,
                left: 8,
                right: 8,
              }),
            )}
          />
        </div>
        <div className={`${rootPrefixCls}--card__info`}>
          <span className={`${rootPrefixCls}--card__info-icon`}>{iconRender(file)}</span>

          <div className={`${rootPrefixCls}--card__info-name`}>
            <FileName file={file} nameRenderer={nameRenderer} />
          </div>
          <span
            className={clsx(
              `${rootPrefixCls}--card__info-operations`,
              stylex(stylex.create({ flexGrow: 0, flexShrink: 0 })),
            )}
          >
            {errorNode}
            {previewNode}
            {downloadNode}
            {retryNode}
            {removeNode}
          </span>
        </div>
      </div>
    </FileContextProvider>
  );
};
