// @universe-design/react/src/upload/uploadFile/index.tsx
import * as React from 'react';
import { CardFile } from './CardFile';
import { PictureFile } from './PictureFile';
import type { TextFileProps } from './TextFile';
import { TextFile, TextFileAction } from './TextFile';
import { useUploadFile } from './useUploadFile';
import type { CardFileProps } from './CardFile';
import type { PictureFileProps } from './PictureFile';

export const UploadTextFile = (props: TextFileProps) => {
  const uploadFileProps = useUploadFile(props);
  return <TextFile {...uploadFileProps} />;
};

export type SimpleTextFileProps = Omit<TextFileProps, 'isSimple'>;
const SimpleTextFile = (props: SimpleTextFileProps) => <TextFile {...props} isSimple />;

export const UploadSimpleTextFile = (props: SimpleTextFileProps) => {
  const uploadFileProps = useUploadFile(props);
  return <SimpleTextFile {...uploadFileProps} />;
};

export const UploadCardFile = (props: CardFileProps) => {
  const uploadFileProps = useUploadFile(props);
  return <CardFile {...uploadFileProps} />;
};

export const UploadPictureFile = (props: PictureFileProps) => {
  const uploadFileProps = useUploadFile(props);
  return <PictureFile {...uploadFileProps} />;
};
export { TextFile, TextFileAction, CardFile, PictureFile };
