// @universe-design/react/src/upload/uploadFile/interface.ts
import type * as React from 'react';

import type { UseFileProps } from './useFile';
import type { StyleProps } from '../../util';
import type { ProgressProps } from '../../progress';
import type { RcFile } from '../interface';
import type { TextFileProps } from './TextFile';
import type { PictureFileProps } from './PictureFile';
import type { CardFileProps } from './CardFile';
/**
 * @public
 * `<UploadTextFile>`、`<UploadCardFile>`、`<UploadSimpleTextFile>`、`<UploadPictureFile>` 等共有的属性。
 * @docgen
 *
 */
export type FileProps = UseFileProps &
  StyleProps & {
    /**
     * 自定义文件名渲染
     */
    nameRenderer?: (file: UploadFile) => React.ReactElement;
  };
/**
 * 上传进度条样式。可参考<Progress /> 组件。
 * @public
 */
export type UploadProgressProps = Omit<ProgressProps, 'percent' | 'type'>;

/**
 * @public
 * 自定义文件预览行为，需要返回文件 url
 */
export type PreviewFileHandler = (file: File | Blob) => PromiseLike<string>;

export type OperationIcon = React.ReactNode | ((file: UploadFile) => React.ReactNode);

/**
 * 所有的文件上传状态。
 * @public
 */
export type UploadFileStatus =
  | 'error'
  | 'waiting'
  | 'success'
  | 'uploading'
  | 'downloading'
  | 'downloaded'
  | 'downloadFailed'
  | 'removed';

/**
 * 文件列表每一项的数据结构。`Response` 为上传接口 response 的类型。
 * @public
 * @typeParam Response - 上传接口的返回值类型
 */
export interface UploadFile<Response = any> {
  /**
   * 文件唯一表示，不指定的话，将会自动生成
   */
  uid?: string;
  /**
   * 文件大小
   */
  size?: number;
  /**
   * 文件名
   */
  name: string;
  /**
   * @deprecated 无用处
   */
  fileName?: string;
  /**
   * @deprecated 无用处
   */
  lastModified?: number;
  /**
   * @deprecated 无用处
   */
  lastModifiedDate?: Date;
  /**
   * 下载地址
   */
  url?: string;
  /**
   * 上传状态，不同状态展示颜色也会有所不同
   */
  status?: UploadFileStatus;
  /**
   * 当前文件上传进度，
   */
  percent?: number;
  /**
   * 已上传（或已下载）数据大小
   */
  loaded?: number;
  /**
   * 缩略图地址
   */
  thumbUrl?: string;
  /**
   * @internal 无用处
   */
  originFileObj?: RcFile;
  /**
   * 上传接口响应数据
   */
  response?: Response;
  /**
   * 上传过程中抛出的异常
   */
  error?: any;
  /**
   * @internal 无用处
   */
  linkProps?: any;
  /**
   * @internal 无用处
   */
  type?: string;
  /**
   * 该文件上传时 创建的XMLHttpRequest
   */
  xhr?: unknown;
  /**
   * 显示在上传列表的失败原因，默认为 "Error"
   * @defaultValue "Error"
   */
  failedReason?: string;
}

export type { TextFileProps, UseFileProps, PictureFileProps, CardFileProps };
