// @universe-design/react/src/upload/uploadFile/TextFile.tsx
/** @jsxRuntime classic */
/** @jsx jsx */
/** @jsxFrag React.Fragment */
import clsx from 'clsx';
import React from 'react';
import { useLocaleReceiver } from '../../config-provider';
import { TextOverflow } from '../../shared/components/text-overflow';
import { UploadProgress } from './UploadProgress';
import { formatFileSize, isError } from '../utils';
import { useFile } from './useFile';
import { FileName } from './FileName';
import type { FileProps, UploadFile } from './interface';
import { FileContextProvider, useFileContext } from './FileContext';
import stylex from '@universe-design/stylex';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx } from '@universe-design/stylex/react';
import '../style/text-file.less';

const DownloadAction = () => {
  const { downloadNode } = useFileContext();
  return downloadNode;
};

const RetryAction = () => {
  const { retryNode } = useFileContext();
  return retryNode;
};

const PreviewAction = () => {
  const { previewNode } = useFileContext();
  return previewNode;
};

const RemoveAction = () => {
  const { removeNode } = useFileContext();
  return removeNode;
};

export const TextFileAction = {
  Download: DownloadAction,
  Retry: RetryAction,
  Preview: PreviewAction,
  Remove: RemoveAction,
};

/**
 * 与 Upload 无关的 File 组件
 */
export interface TextFileProps extends FileProps {
  /**
   * 是否是简略模式
   */
  isSimple?: boolean;
  /**
   * 自定义操作按钮，如果设置则不显示内置按钮
   */
  action?: React.ReactNode;
}

const FileStatus = ({ file, prefixCls }: { file: UploadFile; prefixCls: string }) => {
  const locale = useLocaleReceiver('upload');
  if (file.status === 'waiting') {
    return (
      <TextOverflow as="p" className={`${prefixCls}-info-subtitle`} lineClamp={1}>
        {locale.waiting}
      </TextOverflow>
    );
  }
  if (isError(file)) {
    const defaultFailedReason = file.status === 'error' ? locale.failed : locale.downloadFailed;
    const failedReason = file.failedReason || defaultFailedReason;
    return (
      <TextOverflow as="p" className={`${prefixCls}-info-error`} lineClamp={2}>
        {failedReason}
      </TextOverflow>
    );
  }
  if (file.status === 'uploading' || file.status === 'downloading') {
    return (
      <TextOverflow as="p" className={`${prefixCls}-info-subtitle`} lineClamp={1}>
        {file.status === 'downloading' && `${locale.downloading} `}
        {formatFileSize(file.loaded, locale)} / {formatFileSize(file.size, locale)}
      </TextOverflow>
    );
  }
  return (
    <TextOverflow as="p" className={`${prefixCls}-info-subtitle`} lineClamp={1}>
      {formatFileSize(file.size, locale)}
    </TextOverflow>
  );
};

const defaultAction = (
  <>
    <TextFileAction.Preview key="preview" />
    <TextFileAction.Retry key="retry" />
    <TextFileAction.Download key="download" />
    <TextFileAction.Remove key="remove" />
  </>
);
export function TextFile(props: TextFileProps) {
  const { style, isSimple, nameRenderer, action = defaultAction } = props;

  const fileCtxValue = useFile(props);
  const { file, errorNode, iconRender, rootPrefixCls, onPreview, progressProps, classNames } = fileCtxValue;

  const operationAreaRef = React.useRef<HTMLDivElement>(null);
  const previewable = ['success', 'downloading', 'downloaded', 'downloadFailed'].includes(file.status);
  const previewFile: React.MouseEventHandler<HTMLDivElement> = function (e) {
    if (!previewable) {
      return;
    }

    const { target } = e;
    // 防止操作按钮触发 item preview
    if (
      (target instanceof HTMLElement || target instanceof SVGElement) &&
      (e.target === operationAreaRef.current || operationAreaRef.current?.contains(target))
    ) {
      return;
    }
    onPreview?.();
  };
  return (
    <FileContextProvider value={fileCtxValue}>
      <div
        className={clsx(classNames, isSimple ? `${rootPrefixCls}--simple-text` : `${rootPrefixCls}--text`)}
        onClick={previewFile}
        style={style}
        sx={[previewable && stylex.create({ cursor: 'pointer' })]}
      >
        <div className={`${rootPrefixCls}-container`}>
          <div className={`${rootPrefixCls}-icon`}>{iconRender?.(file)}</div>
          <div className={`${rootPrefixCls}-info`}>
            <FileName file={file} nameRenderer={nameRenderer} className={`${rootPrefixCls}-info-name`} />
            {!isSimple && <FileStatus file={file} prefixCls={rootPrefixCls} />}
          </div>
          <div className={`${rootPrefixCls}-operations`} ref={operationAreaRef}>
            {isSimple && errorNode}
            {action}
          </div>
        </div>
        <UploadProgress file={file} progressProps={progressProps} prefixCls={rootPrefixCls} />
      </div>
    </FileContextProvider>
  );
}
