// @universe-design/react/src/upload/uploadFile/useFile.tsx
import * as React from 'react';
import DeleteTrashOutlined from '@universe-design/icons-react/DeleteTrashOutlined';
import DownloadOutlined from '@universe-design/icons-react/DownloadOutlined';
import RefreshOutlined from '@universe-design/icons-react/RefreshOutlined';
import VisibleOutlined from '@universe-design/icons-react/VisibleOutlined';
import WarningFilled from '@universe-design/icons-react/WarningFilled';
import stylex from '@universe-design/stylex';
import clsx from 'clsx';
import { useCssScope, useLocaleReceiver } from '../../config-provider/hooks';
import { internalPreview, internalDownload, isError } from '../utils';
import { Tooltip } from '../../tooltip';
import type { IconButtonProps } from '../../button';
import { IconButton } from '../../button';

import type { StyleProps } from '../../util';
import type { OperationIcon, UploadProgressProps } from './interface';
import type { UploadFile } from './interface';
import type { UploadListType } from '../uploadList/interface';
import getFileIcon from '../utils/getFileIcon';
import { useFileContext } from './FileContext';

export type UseFileProps = {
  /**
   * 是否处于禁用态
   */
  disabled?: boolean;
  /**
   * 文件
   */
  file: UploadFile;
  /**
   * 进度条属性
   */
  progress?: UploadProgressProps;
  /**
   * 自定义 preview 逻辑
   */
  onPreview?: () => void;
  /**
   * 点击下载文件时的回调，如果没有指定，则默认跳转到文件 url 对应的标签页
   */
  onDownload?: () => void;
  /**
   * 点击重试按钮的回调
   */
  onRetry?: () => void | boolean;
  /**
   * 点击移除文件时的回调，返回值为 false 时不移除。支持返回一个 Promise 对象，Promise 对象 resolve(false) 或 reject 时不移除
   */
  onRemove?: () => void | boolean | Promise<void | boolean>;
  /**
   * 自定义文件显示何种图标，返回 `null` 则会用内置图标兜底
   */
  iconRender?: (file: UploadFile, listType?: UploadListType) => React.ReactNode;
  prefixCls?: string;
  /**
   * 是否显示删除按钮
   */
  showRemoveIcon?: boolean;
  /**
   * 是否显示预览按钮
   */
  showPreviewIcon?: boolean;
  /**
   * 是否显示下载按钮
   */
  showDownloadIcon?: boolean;
  /**
   * 是否显示重试按钮
   */
  showRetryIcon?: boolean;
  /**
   * 是否显示上传失败 Icon，图片墙（PictureFile）场景下仅展示 tooltip
   */
  showErrorIcon?: boolean;
  /**
   * 自定义删除按钮
   */
  removeIcon?: OperationIcon;
  /**
   * 自定义下载按钮
   */
  downloadIcon?: OperationIcon;
  /**
   * 自定义预览按钮
   */
  previewIcon?: OperationIcon;
  /**
   * 自定义重试按钮
   */
  retryIcon?: OperationIcon;
  /**
   * 自定义上传失败 Icon
   */
  errorIcon?: OperationIcon;
} & StyleProps;

export type InternalItemProps = UseFileProps;

export function OperationIconComp({
  icon,
  handleClick,
  tooltip,
  className,
  style,
  color,
}: {
  icon: OperationIcon;
  handleClick?: (file: UploadFile) => void;
  tooltip?: string;
} & StyleProps &
  Pick<IconButtonProps, 'color'>) {
  const { file, rootPrefixCls: prefixCls } = useFileContext();
  if (typeof icon === 'function') {
    icon = icon(file);
  }

  if (!icon) {
    return null;
  }
  return (
    <Tooltip title={tooltip}>
      <IconButton
        className={clsx(`${prefixCls}-operation-icon`, className)}
        icon={icon}
        onClick={() => handleClick?.(file)}
        style={style}
        size="small"
        color={color}
      />
    </Tooltip>
  );
}

export function useFile(props: UseFileProps) {
  const { file: fileProp, prefixCls, className } = props;

  const file = React.useMemo(
    () => ({
      status: 'success' as const,
      ...fileProp,
    }),
    [fileProp],
  );

  const defaultFileProps = React.useMemo(
    () => ({
      disabled: false,
      progress: {
        strokeWidth: 4,
        showInfo: false,
      },
      showRemoveIcon: true,
      showDownloadIcon: true,
      showPreviewIcon: true,
      showErrorIcon: true,
      showRetryIcon: true,
      removeIcon: <DeleteTrashOutlined data-testid="remove-icon" />,
      onPreview: () => internalPreview(file),
      previewIcon: <VisibleOutlined data-testid="preview-icon" />,
      onDownload: () => internalDownload(file),
      downloadIcon: <DownloadOutlined data-testid="download-icon" />,
      retryIcon: <RefreshOutlined data-testid="retry-icon" />,
      errorIcon: (
        <WarningFilled
          data-testid="error-icon"
          className={stylex(stylex.create({ color: $token['function-danger-content-default'] }))}
        />
      ),
    }),
    [file],
  );

  const mergedFileProps = {
    // 默认值优先级最低
    ...defaultFileProps,
    // props 优先级最高
    ...props,
  };

  const {
    showDownloadIcon,
    downloadIcon,
    showErrorIcon,
    showPreviewIcon,
    showRemoveIcon,
    showRetryIcon,
    errorIcon,
    previewIcon,
    progress,
    disabled,
    removeIcon,
    retryIcon,
    onRemove,
    onRetry,
    onDownload,
    onPreview,
    iconRender: iconRenderProp,
  } = mergedFileProps;

  const rootPrefixCls = useCssScope('upload-item', prefixCls)();

  const locale = useLocaleReceiver('upload');
  const status = file.status || 'success';

  const downloadNode = React.useMemo(() => {
    if (!showDownloadIcon || status !== 'success') {
      return null;
    }

    return <OperationIconComp icon={downloadIcon} handleClick={onDownload} tooltip={locale.download} />;
  }, [downloadIcon, locale.download, onDownload, showDownloadIcon, status]);

  const errorNode = React.useMemo(() => {
    if (!showErrorIcon || status !== 'error') {
      return null;
    }
    const failedReason = file.failedReason || locale.failed;
    return <OperationIconComp icon={errorIcon} tooltip={failedReason} />;
  }, [errorIcon, file, locale.failed, showErrorIcon, status]);

  const previewable = showPreviewIcon && ['success', 'downloaded'].includes(status);

  const previewNode = React.useMemo(() => {
    if (!previewable || !previewIcon) {
      return null;
    }
    return <OperationIconComp icon={previewIcon} tooltip={locale.preview} handleClick={onPreview} />;
  }, [locale.preview, onPreview, previewable, previewIcon]);

  const retryNode = React.useMemo(() => {
    if (!showRetryIcon || status !== 'error' || disabled || !retryIcon) {
      return null;
    }
    return (
      <OperationIconComp
        icon={retryIcon}
        tooltip={locale.uploadAgain}
        handleClick={onRetry}
        className={`${rootPrefixCls}-operation-retry`}
        color="primary"
      />
    );
  }, [disabled, locale.uploadAgain, onRetry, retryIcon, showRetryIcon, status, rootPrefixCls]);

  const removeNode = React.useMemo(() => {
    if (!showRemoveIcon || disabled || file.status === 'downloading') {
      return null;
    }
    return <OperationIconComp icon={removeIcon} tooltip={locale.delete} handleClick={onRemove} />;
  }, [disabled, file, locale.delete, onRemove, removeIcon, showRemoveIcon]);

  function iconRender(f: UploadFile): React.ReactNode {
    return iconRenderProp?.(f) ?? getFileIcon(f);
  }

  const classNames = clsx(
    className,
    rootPrefixCls,
    `${rootPrefixCls}--${file.status}`,
    isError(file) ? `${rootPrefixCls}--error` : `${rootPrefixCls}--normal`,
  );

  return {
    downloadNode,
    errorNode,
    removeNode,
    previewNode,
    retryNode,
    showPreviewIcon,
    showErrorIcon,
    iconRender,
    onRemove,
    onPreview,
    file,
    disabled,
    rootPrefixCls,
    progressProps: progress,
    classNames,
  };
}

export type FileItemContextType = ReturnType<typeof useFile>;
