// @universe-design/react/src/upload/uploadFile/useUploadFile.ts
import React from 'react';
import { useUploadContext } from '../hooks/UploadContext';
import { internalPreview, internalDownload } from '../utils';
import type { FileProps } from './interface';

export const useUploadFile = (props: FileProps) => {
  const {
    file,
    onRemove: onRemoveProp,
    onRetry: onRetryProp,
    onPreview: onPreviewProp,
    onDownload: onDownloadProp,
    disabled: disabledProp,
    ...resetProps
  } = props;
  const onPreview = React.useCallback(
    () => (onPreviewProp ? onPreviewProp() : internalPreview(file)),
    [file, onPreviewProp],
  );
  const onDownload = React.useCallback(
    () => (onDownloadProp ? onDownloadProp() : internalDownload(file)),
    [file, onDownloadProp],
  );

  const { onRetry: onRetryCtx, disabled, onRemove: onRemoveCtx } = useUploadContext();

  const onRemove = React.useCallback(() => {
    Promise.resolve(typeof onRemoveProp === 'function' ? onRemoveProp() : true).then(ret => {
      // Prevent removing file
      if (ret === false) {
        return;
      }
      onRemoveCtx(file);
    });
  }, [file, onRemoveCtx, onRemoveProp]);

  const onRetry = React.useCallback(() => {
    onRetryCtx(file);
    onRetryProp?.();
  }, [file, onRetryCtx, onRetryProp]);

  return {
    onPreview,
    onRemove,
    onRetry,
    onDownload,
    disabled: disabledProp ?? disabled,
    file,
    ...resetProps,
  };
};
