// @universe-design/react/es/shared/components/popper/test-utils/components.js
import _extends from "@babel/runtime/helpers/esm/extends";
import * as React from 'react';
import { Popper, PopperArrow } from '@universe-design/react';
import '../index.stories.less';
export const buttonWidth = 100;
const buttonHeight = 50;
const timeout = {
  appear: 1,
  enter: 1,
  exit: 1
};

// eslint-disable-next-line max-len
const longText = "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.";
export const Popover = props => {
  const {
    style,
    referenceProps: {
      style: referenceStyle,
      ...referenceProps
    } = {},
    placement,
    hasArrow = true
  } = props;
  return /*#__PURE__*/React.createElement(Popper, _extends({
    timeout: timeout,
    popup: /*#__PURE__*/React.createElement("div", {
      style: {
        boxSizing: 'border-box',
        flex: 'auto',
        minWidth: 0,
        minHeight: 0,
        overflow: 'auto',
        color: 'white',
        background: 'rgba(255, 0, 0, 0.3)',
        maxWidth: 200,
        height: 200,
        border: '2px solid rgba(0, 255, 0, 0.3)',
        ...style
      }
    }, longText),
    trigger: "click",
    defaultVisible: true
  }, hasArrow ? {
    arrowPadding: 8,
    offset: 10,
    arrow: /*#__PURE__*/React.createElement(PopperArrow, {
      hasStroke: true
    })
  } : {
    arrowPadding: 0,
    offset: 4
  }, props), /*#__PURE__*/React.createElement("button", _extends({
    type: "button",
    className: "reference",
    style: {
      width: buttonWidth,
      height: buttonHeight,
      whiteSpace: 'nowrap',
      overflow: 'hidden',
      textOverflow: 'ellipsis',
      ...referenceStyle
    }
    // eslint-disable-next-line react/no-children-prop
    ,
    children: placement
  }, referenceProps)));
};
