// @universe-design/styles/esm/ccm-ccm-chat/index.light.atom.js
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
export const ccmtoken_chart_aurora_b100 = '#bdd2ff';
export const ccmtoken_chart_aurora_b200 = '#8fb4ff';
export const ccmtoken_chart_aurora_b300 = '#4c88ff';
export const ccmtoken_chart_aurora_b400 = '#3370eb';
export const ccmtoken_chart_aurora_b500 = '#2655b6';
export const ccmtoken_chart_aurora_b600 = '#194294';
export const ccmtoken_chart_aurora_c100 = '#ffc2e5';
export const ccmtoken_chart_aurora_c200 = '#fc94cf';
export const ccmtoken_chart_aurora_c300 = '#ed77ba';
export const ccmtoken_chart_aurora_c400 = '#c24a8e';
export const ccmtoken_chart_aurora_c500 = '#94386c';
export const ccmtoken_chart_aurora_c600 = '#782b57';
export const ccmtoken_chart_aurora_g100 = '#99e490';
export const ccmtoken_chart_aurora_g200 = '#69cc5c';
export const ccmtoken_chart_aurora_g300 = '#51ba43';
export const ccmtoken_chart_aurora_g400 = '#419e34';
export const ccmtoken_chart_aurora_g500 = '#2f7526';
export const ccmtoken_chart_aurora_g600 = '#21511a';
export const ccmtoken_chart_aurora_o100 = '#fec88b';
export const ccmtoken_chart_aurora_o200 = '#f89e44';
export const ccmtoken_chart_aurora_o300 = '#f3871b';
export const ccmtoken_chart_aurora_o400 = '#db7018';
export const ccmtoken_chart_aurora_o500 = '#a15317';
export const ccmtoken_chart_aurora_o600 = '#683a12';
export const ccmtoken_chart_aurora_p100 = '#dbc8fd';
export const ccmtoken_chart_aurora_p200 = '#c5a3ff';
export const ccmtoken_chart_aurora_p300 = '#a575fa';
export const ccmtoken_chart_aurora_p400 = '#8c55ec';
export const ccmtoken_chart_aurora_p500 = '#6c39c6';
export const ccmtoken_chart_aurora_p600 = '#5529a3';
export const ccmtoken_chart_aurora_r100 = '#f89e9b';
export const ccmtoken_chart_aurora_r200 = '#f6827e';
export const ccmtoken_chart_aurora_r300 = '#f05b56';
export const ccmtoken_chart_aurora_r400 = '#d14642';
export const ccmtoken_chart_aurora_r500 = '#a03331';
export const ccmtoken_chart_aurora_r600 = '#a03331';
export const ccmtoken_chart_aurora_t100 = '#6fe8d8';
export const ccmtoken_chart_aurora_t200 = '#33d6c0';
export const ccmtoken_chart_aurora_t300 = '#2dbeab';
export const ccmtoken_chart_aurora_t400 = '#10a893';
export const ccmtoken_chart_aurora_t500 = '#0f8575';
export const ccmtoken_chart_aurora_t600 = '#1d4e47';
export const ccmtoken_chart_aurora_w100 = '#89dffe';
export const ccmtoken_chart_aurora_w200 = '#51c3f0';
export const ccmtoken_chart_aurora_w300 = '#25b2e5';
export const ccmtoken_chart_aurora_w400 = '#1099cc';
export const ccmtoken_chart_aurora_w500 = '#146c91';
export const ccmtoken_chart_aurora_w600 = '#164359';
export const ccmtoken_chart_aurora_y100 = '#ffde75';
export const ccmtoken_chart_aurora_y200 = '#fbcb46';
export const ccmtoken_chart_aurora_y300 = '#f0b622';
export const ccmtoken_chart_aurora_y400 = '#d99904';
export const ccmtoken_chart_aurora_y500 = '#ad7a03';
export const ccmtoken_chart_aurora_y600 = '#6f4a01';
