// @universe-design/react/src/shared/components/popper/test-utils/components.tsx
import * as React from 'react';
import type { PopperProps } from '@universe-design/react';
import { Popper, PopperArrow } from '@universe-design/react';
import '../index.stories.less';

export const buttonWidth = 100;
const buttonHeight = 50;

const timeout = { appear: 1, enter: 1, exit: 1 };

// eslint-disable-next-line max-len
const longText =
  "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.";

export type DemoPopoverProps = Omit<PopperProps, 'popup' | 'children'> & {
  referenceProps?: React.ComponentPropsWithoutRef<'button'>;
  hasArrow?: boolean;
};

export const Popover = (props: DemoPopoverProps) => {
  const {
    style,
    referenceProps: { style: referenceStyle, ...referenceProps } = {},
    placement,
    hasArrow = true,
  } = props;
  return (
    <Popper
      timeout={timeout}
      popup={
        <div
          style={{
            boxSizing: 'border-box',
            flex: 'auto',
            minWidth: 0,
            minHeight: 0,
            overflow: 'auto',
            color: 'white',
            background: 'rgba(255, 0, 0, 0.3)',
            maxWidth: 200,
            height: 200,
            border: '2px solid rgba(0, 255, 0, 0.3)',
            ...style,
          }}
        >
          {longText}
        </div>
      }
      trigger="click"
      defaultVisible
      {...(hasArrow
        ? {
            arrowPadding: 8,
            offset: 10,
            arrow: <PopperArrow hasStroke />,
          }
        : {
            arrowPadding: 0,
            offset: 4,
          })}
      {...props}
    >
      <button
        type="button"
        className="reference"
        style={{
          width: buttonWidth,
          height: buttonHeight,
          whiteSpace: 'nowrap',
          overflow: 'hidden',
          textOverflow: 'ellipsis',
          ...referenceStyle,
        }}
        // eslint-disable-next-line react/no-children-prop
        children={placement}
        {...referenceProps}
      />
    </Popper>
  );
};
