// @universe-design/react/lib/react-mini-form/internal/FormInternalContext.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FormInternalContextProvider = void 0;
exports.useFormInternalContext = useFormInternalContext;
exports.useFormInternalContextSelector = useFormInternalContextSelector;
var _useContextSelector = require("use-context-selector");
var _reactUtilities = require("@universe-design/react-utilities");
/**
 * @docgen
 * @typeParam Values - values of this form
 */

const {
  Provider: FormInternalContextProvider,
  useContext: _useFormInternalContext,
  useContextSelector: _useFormInternalContextSelector
} = (0, _reactUtilities.createContextHook)('MiniFormContext', {
  strict: true,
  customizedContextMethods: {
    createContext: _useContextSelector.createContext,
    useContext: _useContextSelector.useContext,
    useContextSelector: _useContextSelector.useContextSelector
  }
});
exports.FormInternalContextProvider = FormInternalContextProvider;
/**
 * useFormInternalContextSelector
 * get form context, similar to redux selector
 */
function useFormInternalContextSelector(selector) {
  return _useFormInternalContextSelector(selector);
}

/**
 * useFormInternalContext
 * get from context
 */
function useFormInternalContext() {
  return _useFormInternalContext();
}
