// @universe-design/react/lib/react-mini-form/internal/useFormInternal.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useForm = useForm;
var _get = _interopRequireDefault(require("lodash/get"));
var _set = _interopRequireDefault(require("lodash/set"));
var _pick = _interopRequireDefault(require("lodash/pick"));
var _react = require("react");
var _rxjs = require("rxjs");
var _operators = require("rxjs/operators");
var _reactUtilities = require("@universe-design/react-utilities");
var _hooks = require("../../hooks");
var _util = require("../../util");
var _shared = require("../shared");
var _formImmer = require("./formImmer");
/* eslint-disable max-depth */
// nolint: long_function

function resolveStateActionsValue(prevValues, actionsValue) {
  const resolvedValues = (0, _util.isFunction)(actionsValue) ? actionsValue(prevValues) : actionsValue;
  return resolvedValues;
}

/**
 * Form State: 1. values; 2. errors; 3. toucheds
 * FromItem change 触发 setValues 并且触发 validate
 *
 * validate 包含两个个环节: 1. 所有表单项级别 validate; 2. 表单级别 validate
 * 表单项 validate 只对已经 touched 和已经注册的字段进行校验
 */
function useForm(props) {
  /**
     * FIXME use event callback may cause problems in some rare cases. Assume that you have a form and a single input, now you want to do some check for the input value and form value in `onSubmit` callback. Assume that you have a long-time async validation, and you change the single input value to `abc` from empty during validation. If use useCallback, the input value in  `onSubmit`'s closure will be `""`, but when using useEventCallback, the input value will be `"abc"`
     * @example
     * ```
      function delayReturn<T>(value: T, timeout?: number) {
        return new Promise<T>((resolve) => {
          window.setTimeout(() => {
            resolve(value);
          }, timeout);
        });
      }
       function A() {
        const [value, setValue] = useState<string>("");
        return (
          <>
            <input value={value} onChange={(e) => setValue(e.target.value)} />
            <Form
              initialValues={initialValues}
              onSubmit={() => {
                alert(value);
              }}
            >
              <input type={"submit"} />
              <Field name={["input"]} validate={() => delayReturn(void 0, 5000)}>
                <input />
              </Field>
            </Form>
          </>
        );
      }
     * ```
     */
  const onSubmit = (0, _hooks.useEventCallback)(props.onSubmit);
  const validate = (0, _hooks.useEventCallback)(props.validate);
  const onValuesChange = (0, _hooks.useEventCallback)(props.onValuesChange);
  const onSubmitError = (0, _hooks.useEventCallback)(props.onSubmitError);
  const {
    initialValues,
    debug = false
  } = props;

  // ================ From Context ================

  // form 注册中心，用于记录 form item 的 name validator 等，使用 field id 作为 key
  const registry = (0, _reactUtilities.useConstant)(() => new Map());
  const serializedNameRegistryValueIndexes = (0, _reactUtilities.useConstant)(() => new Map());

  // values
  const [values, _setStateValues] = (0, _react.useState)((0, _formImmer.freeze)(initialValues));
  const valuesRef = (0, _react.useRef)(values);
  const setStateValues = (0, _react.useCallback)((nextValues, changedConfig) => {
    const nextV = (0, _formImmer.freeze)(resolveStateActionsValue(valuesRef.current, nextValues));
    valuesRef.current = nextV;
    _setStateValues(nextV);
    if (changedConfig !== null && changedConfig !== void 0 && changedConfig.triggerByUser && (0, _util.isFunction)(onValuesChange)) {
      onValuesChange((0, _pick.default)(nextV, [(0, _shared.unmarshalNamePath)(changedConfig.changedNamePath)]), nextV);
    }
  }, [onValuesChange]);

  // errors
  const [errors, _setStateErrors] = (0, _react.useState)();
  const errorsRef = (0, _react.useRef)(errors);
  const setStateErrors = (0, _react.useCallback)(nextValues => {
    const nextV = (0, _formImmer.freeze)(resolveStateActionsValue(errorsRef.current, nextValues));
    errorsRef.current = nextV;
    _setStateErrors(nextV);
  }, []);

  // touched
  // latest value store，not a state store
  const touchedRef = (0, _react.useRef)({});
  const _setFieldTouched = (0, _react.useCallback)((serializedName, _touched) => {
    (0, _set.default)(touchedRef.current, (0, _shared.unmarshalNamePath)(serializedName), _touched);
  }, []);
  const _setFieldTouchedState = (0, _react.useCallback)((serializedName, _touched) => {
    (0, _set.default)(touchedRef.current, (0, _shared.unmarshalNamePath)(serializedName), _touched);
  }, []);
  debug && !_util.__PROD__ &&
  // eslint-disable-next-line react-hooks/rules-of-hooks
  (0, _reactUtilities.useLogger)('react-mini-form', {
    registry,
    values,
    errors,
    touched: touchedRef.current
  });

  /**
   * ================ Validation ================
   */

  // isValidating
  const [isValidating, setIsValidating] = (0, _react.useState)(false);
  // validate 订阅中心
  const validate$ = (0, _reactUtilities.useConstant)(() => new _rxjs.Subject());
  const _validateSingleFieldLevel$ = (0, _react.useCallback)(function (serializedName, event) {
    var _serializedNameRegist, _serializedNameRegist2;
    const validators = (_serializedNameRegist = (_serializedNameRegist2 = serializedNameRegistryValueIndexes.get(serializedName)) === null || _serializedNameRegist2 === void 0 ? void 0 : _serializedNameRegist2.validate.current) !== null && _serializedNameRegist !== void 0 ? _serializedNameRegist : [];
    const filteredValidators = validators.filter(v =>
    // if name is not undefined, this is triggered by onChange, so we should only validate this items
    (event.serializedName === undefined || event.serializedName === serializedName && (0, _get.default)(touchedRef.current, (0, _shared.unmarshalNamePath)(serializedName))) && (
    // namePaths from validateFields contains current field
    event.namePaths === undefined || event.namePaths.some(p => (0, _shared.marshalNamePath)(p) === serializedName)) && (
    // current field's validateTrigger contains trigger event
    v.validateTrigger.includes(_shared.DEFAULT_VALIDATE_TRIGGER) || event.triggerName === _shared.DEFAULT_VALIDATE_TRIGGER || v.validateTrigger.includes(event.triggerName)));

    // 如果没有符合条件的 validators 就标记为跳过，后面就不处理 combined errors 逻辑了
    if (!filteredValidators.length) {
      return (0, _rxjs.of)({
        errorMessage: undefined,
        skipped: true
      });
    }
    const validatePromise = new Promise(async resolve => {
      for (const validator of filteredValidators) {
        const maybeErrorMessage = await validator.validate(event.value, (0, _shared.unmarshalNamePath)(serializedName));
        if (maybeErrorMessage) {
          resolve({
            errorMessage: maybeErrorMessage,
            skipped: false
          });
          return;
        }
      }
      resolve({
        errorMessage: undefined,
        skipped: false
      });
    });
    return (0, _rxjs.from)(validatePromise);
  }, [serializedNameRegistryValueIndexes]);

  // mounted
  const isMounted = (0, _hooks.useMountedState)();
  const _validateMultipleFieldLevel$ = (0, _react.useCallback)(event =>
  // TODO TO BE CHECKED!! Please pay attention, we will do all field level validation for now.
  // if (name) {
  //   const k = marshalNamePath(name);
  //   const v = registry.get(k);
  //   if (v) {
  //     nameEntries.push([k, v]);
  //   }
  // } else {
  //   nameEntries = [...registry.entries()];
  // }
  // console.log('_validateMultipleFieldLevel$', nameEntries);
  (0, _rxjs.forkJoin)(
  // TODO: field 的每个 error 互不关联，不需要等待全部返回
  [...registry.values()].map(v => {
    const value = (0, _get.default)(valuesRef.current, (0, _shared.unmarshalNamePath)(v.serializedName));
    return _validateSingleFieldLevel$(v.serializedName, {
      value,
      ...event
    });
  })).pipe((0, _operators.map)(_errors => [...registry.values()].reduce((pre, v, i) => {
    const error = _errors[i];
    const name = (0, _shared.unmarshalNamePath)(v.serializedName);
    if (error.skipped) {
      (0, _set.default)(pre, name, (0, _get.default)(errorsRef.current, name));
    } else {
      (0, _set.default)(pre, name, error.errorMessage);
    }
    return pre;
  }, {}))), [_validateSingleFieldLevel$, registry]);

  /**
   * _validateMultipleFieldLevel
   * @param values current values state
   * @param name if undefined, validate all fields, if assigned, validate name.
   */
  const _validateFormLevel$ = (0, _react.useCallback)(name => {
    try {
      if (!(0, _util.isFunction)(validate)) {
        return (0, _rxjs.of)(undefined);
      }
      const maybePromise = validate(valuesRef.current, name);
      if (!maybePromise) {
        return (0, _rxjs.of)(undefined);
      }
      if ((0, _util.isPromise)(maybePromise)) {
        return (0, _rxjs.from)(maybePromise);
      }
      return (0, _rxjs.of)(maybePromise);
    } catch (err) {
      // sync error to observable
      return (0, _rxjs.throwError)(err);
    }
  }, [validate]);
  const _validateAll$ = (0, _react.useCallback)(event => (0, _rxjs.forkJoin)([_validateMultipleFieldLevel$(event), _validateFormLevel$(event.serializedName)]).pipe((0, _operators.map)(_ref => {
    let [fieldErrors, formErrors] = _ref;
    // 这里的 merge 要用 lodash merge，因为我们不能让 formErrors 的 undefined 覆盖 fieldErrors 里面已有的 errors
    const combinedErrors = {};
    registry.forEach(_ref2 => {
      let {
        serializedName: _serializedName
      } = _ref2;
      const name = (0, _shared.unmarshalNamePath)(_serializedName);
      const fieldErrorMessage = (0, _get.default)(fieldErrors, name);
      const formErrorMessage = (0, _get.default)(formErrors, name);
      const errorMessage = fieldErrorMessage || formErrorMessage;
      if (errorMessage) {
        // 如果出现了 error 则认为是 touch 了
        _setFieldTouched(_serializedName, true);
        (0, _set.default)(combinedErrors, name, errorMessage);
      }
    });
    return combinedErrors;
  })), [_validateMultipleFieldLevel$, _validateFormLevel$, registry, _setFieldTouched]);
  (0, _react.useEffect)(() => {
    const sub = validate$.pipe((0, _operators.tap)(() => {
      setIsValidating(true);
    }), (0, _operators.delay)(0), (0, _operators.switchMap)(_ref3 => {
      let {
        serializedName,
        triggerName
      } = _ref3;
      return _validateAll$({
        serializedName,
        triggerName
      });
    })).subscribe({
      next: combinedErrors => {
        if (isMounted()) {
          setIsValidating(false);
          setStateErrors(combinedErrors);
        }
      },
      error: error => {
        (0, _shared.printFormValidationError)(error);
        if (isMounted()) {
          setIsValidating(false);
        }
      },
      complete: () => {
        if (isMounted()) {
          setIsValidating(false);
        }
      }
    });
    return () => {
      sub.unsubscribe();
    };
  }, [_validateAll$, isMounted, setStateErrors, validate$, valuesRef]);
  (0, _hooks.useEffectOnce)(() => () => {
    validate$.complete();
  });

  // TODO rare case
  // 1. change 触发的 validate
  // 2. 修改 value
  // 3. validateFields
  // 4. validateFields 先返回值，change 触发的后返回值
  // 5. 则 errors 有可能会不正确

  const validateFields = (0, _react.useCallback)((nameList, options) => {
    setIsValidating(true);
    return new Promise((resolve, reject) => {
      // 强行在下一次渲染之后进行校验，解决 form item 根据 values 生成场景校验在 form item 注销之前带来的问题。
      setTimeout(() => {
        (0, _rxjs.firstValueFrom)(_validateAll$({
          namePaths: nameList,
          triggerName: (options === null || options === void 0 ? void 0 : options.triggerName) || _shared.DEFAULT_VALIDATE_TRIGGER
        }), {
          defaultValue: undefined
        }).then(combinedErrors => {
          if (isMounted()) {
            setIsValidating(false);
            setStateErrors(combinedErrors);
          }
          resolve(combinedErrors);
        }).catch(err => {
          (0, _shared.printFormValidationError)(err);
          if (isMounted()) {
            setIsValidating(false);
          }
          reject(err);
        });
      });
    });
  }, [_validateAll$, isMounted, setStateErrors]);
  const setValues = (0, _react.useCallback)((_values, shouldValidate) => {
    setStateValues(prevValues => {
      const resolvedValues = (0, _util.isFunction)(_values) ? _values(prevValues) : _values;
      const willValidate = shouldValidate === undefined ? true : shouldValidate;
      if (willValidate) {
        validate$.next({
          triggerName: _shared.DEFAULT_VALIDATE_TRIGGER
        });
      }
      return resolvedValues;
    });
  }, [setStateValues, validate$]);
  const setFieldsValue = (0, _react.useCallback)((partialValues, shouldValidate) => {
    setStateValues(prevValues => {
      const fieldPathSet = new Set([...registry.values()].map(o => o.serializedName));
      // 这里的 merge 要用 mergeWithUndefined，因为用户可能会把某个 field 设置为 undefined，这个时候用户的预期行为其实是要 merge undefined 值的
      const resolvedValues = (0, _formImmer.produce)(prevValues, draft => {
        (0, _shared.mergeWithUndefined)(draft, (0, _util.isFunction)(partialValues) ? partialValues(prevValues) : partialValues, fieldPathSet);
      });
      const willValidate = shouldValidate === undefined ? true : shouldValidate;
      if (willValidate) {
        validate$.next({
          triggerName: _shared.DEFAULT_VALIDATE_TRIGGER
        });
      }
      return resolvedValues;
    });
  }, [registry, setStateValues, validate$]);
  const _setFieldValue = (0, _react.useCallback)((serializedName, value, config) => {
    setStateValues(prevValues => {
      const nextValues = (0, _formImmer.produce)(prevValues, draft => {
        (0, _set.default)(draft, (0, _shared.unmarshalNamePath)(serializedName), (0, _util.isFunction)(value) ? value(prevValues) : value);
      });
      // 若为用户事件触发 onChange，校验逻辑已经 merge 进 trigger 中
      const willValidate = config.shouldValidate === undefined ? !config.triggerByUser : config.shouldValidate;
      if (willValidate) {
        validate$.next({
          serializedName,
          triggerName: config.triggerName || _shared.DEFAULT_VALIDATE_TRIGGER
        });
      }
      return nextValues;
    }, {
      triggerByUser: config.triggerByUser,
      changedNamePath: serializedName
    });
  }, [setStateValues, validate$]);
  const setFieldValue = (0, _react.useCallback)((name, value, shouldValidate) => _setFieldValue((0, _shared.marshalNamePath)(name), value, {
    shouldValidate
  }), [_setFieldValue]);
  const _getStateValues = (0, _react.useCallback)(() => values, [values]);
  const setErrors = (0, _react.useCallback)(_errors => {
    setStateErrors(prevErrors => resolveStateActionsValue(prevErrors, _errors));
  }, [setStateErrors]);
  const resetErrors = (0, _react.useCallback)(() => setErrors(undefined), [setErrors]);
  const setFieldsError = (0, _react.useCallback)(partialErrors => {
    setStateErrors(function () {
      let prevErrors = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      const fieldPathSet = new Set([...registry.values()].map(o => o.serializedName));
      // 这里的 merge 要用 mergeWithUndefined，因为用户可能会把某个 field 设置为 undefined，这个时候用户的预期行为其实是要 merge undefined 值的
      return (0, _formImmer.produce)(prevErrors, draft => {
        (0, _shared.mergeWithUndefined)(draft, resolveStateActionsValue(prevErrors, partialErrors), fieldPathSet);
      });
    });
  }, [registry, setStateErrors]);
  const _setFieldError = (0, _react.useCallback)((name, error) => {
    setStateErrors(function () {
      let prevErrors = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      return (0, _formImmer.produce)(prevErrors, draft => {
        (0, _set.default)(draft, (0, _shared.unmarshalNamePath)(name), resolveStateActionsValue(prevErrors, error));
      });
    });
  }, [setStateErrors]);
  const setFieldError = (0, _react.useCallback)((name, error) => _setFieldError((0, _shared.marshalNamePath)(name), error), [_setFieldError]);

  // ================ Submit ================
  const [isSubmitting, setSubmitting] = (0, _react.useState)(false);
  const checkIsDirty = (0, _react.useCallback)(_errors => Boolean(_errors) && [...registry.values()].some(_ref4 => {
    let {
      serializedName
    } = _ref4;
    return (0, _get.default)(_errors, (0, _shared.unmarshalNamePath)(serializedName));
  }), [registry]);
  const submit = (0, _react.useCallback)(async () => {
    try {
      setSubmitting(true);
      const combinedErrors = await validateFields();
      const isDirty = checkIsDirty(combinedErrors);
      if (isDirty && combinedErrors) {
        (0, _util.isFunction)(onSubmitError) && onSubmitError(combinedErrors);
      } else {
        (0, _util.isFunction)(onSubmit) && (await onSubmit(values));
      }
      if (isMounted()) {
        setSubmitting(false);
      }
    } catch (err) {
      if (isMounted()) {
        setSubmitting(false);
      }
      throw err;
    }
  }, [validateFields, checkIsDirty, isMounted, onSubmitError, onSubmit, values]);
  const reset = (0, _react.useCallback)(() => {
    setStateValues(initialValues);
    setStateErrors({});
    touchedRef.current = {};
  }, [initialValues, setStateErrors, setStateValues]);
  const _registerField = (0, _react.useCallback)(payload => {
    const {
      serializedName
    } = payload;
    if (serializedNameRegistryValueIndexes.get(serializedName)) {
      console.warn(`name: \`${serializedName}\` is duplicated. `);
    }
    registry.set(payload.fieldID, payload);
    serializedNameRegistryValueIndexes.set(serializedName, payload);
  }, [registry, serializedNameRegistryValueIndexes]);
  const _unregisterField = (0, _react.useCallback)(fieldID => {
    var _serializedNameRegist3;
    const registryValue = registry.get(fieldID);
    if (!registryValue) {
      return;
    }
    registry.delete(fieldID);
    /**
     * 「兼容逻辑」
     * 只对 fieldID 和 index registryValue 对得上的情况做 indexes 的清理，因为有可能存在一种情况
     * 首次渲染: list 中只有 index0 的元素
     * 重渲染: list 在 index0 处增加一个新节点，index0 节点的 register 可能会比 index1(之前的 index0) 节点的 unregister 早调用，导致 index0 的 register value 被删除
     */
    if (((_serializedNameRegist3 = serializedNameRegistryValueIndexes.get(registryValue.serializedName)) === null || _serializedNameRegist3 === void 0 ? void 0 : _serializedNameRegist3.fieldID) === fieldID) {
      serializedNameRegistryValueIndexes.delete(registryValue.serializedName);
    }
    _setFieldTouched(registryValue.serializedName, false);
  }, [_setFieldTouched, registry, serializedNameRegistryValueIndexes]);
  return {
    _getStateValues,
    get values() {
      return valuesRef.current;
    },
    get touched() {
      return touchedRef.current;
    },
    setValues,
    setFieldsValue,
    _setFieldValue,
    setFieldValue,
    _setFieldError,
    setErrors,
    resetErrors,
    setFieldsError,
    setFieldError,
    submit,
    reset,
    get errors() {
      return errorsRef.current;
    },
    isSubmitting,
    isValidating,
    get isDirty() {
      return checkIsDirty(errorsRef.current);
    },
    validateFields,
    validate$,
    _setFieldTouched,
    _setFieldTouchedState,
    _registerField,
    _unregisterField
  };
}
