// @universe-design/react/lib/react-mini-form/internal/useFieldInternal.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useFieldContext = useFieldContext;
exports.useFieldInternal = useFieldInternal;
var _get = _interopRequireDefault(require("lodash/get"));
var React = _interopRequireWildcard(require("react"));
var _reactUtilities = require("@universe-design/react-utilities");
var _hooks = require("../../hooks");
var _shared = require("../shared");
var _FormInternalContext = require("./FormInternalContext");
/**
 * @public
 * @docgen
 */

/**
 * @public
 */

/**
 * @public
 * @docgen
 */

/**
 * useFieldContext
 */
function useFieldContext(_ref) {
  let {
    name
  } = _ref;
  const multiLevelPath = (0, _shared.asArray)(name);
  return {
    value: (0, _FormInternalContext.useFormInternalContextSelector)(v => (0, _get.default)(v.values, multiLevelPath)),
    error: (0, _FormInternalContext.useFormInternalContextSelector)(v => v.errors ? (0, _get.default)(v.errors, multiLevelPath) : undefined)
  };
}
let count = 0;
const useFieldID = () => (0, _reactUtilities.useConstant)(() => `field-${count++}`);

/**
 * useField
 */
function useFieldInternal(props) {
  const {
    name,
    validate,
    valuePropName,
    validateTrigger,
    trigger = 'onChange'
  } = props;
  const fieldID = useFieldID();
  const namePathStr = (0, _shared.marshalNamePath)(name);
  const {
    value,
    error
  } = useFieldContext({
    name
  });
  const _setFieldValue = (0, _FormInternalContext.useFormInternalContextSelector)(v => v._setFieldValue);
  const _registerField = (0, _FormInternalContext.useFormInternalContextSelector)(v => v._registerField);
  const _unregisterField = (0, _FormInternalContext.useFormInternalContextSelector)(v => v._unregisterField);
  const _setFieldTouched = (0, _FormInternalContext.useFormInternalContextSelector)(v => v._setFieldTouched);
  const validate$ = (0, _FormInternalContext.useFormInternalContextSelector)(v => v.validate$);
  const forceUpdate = (0, _reactUtilities.useForceUpdate)();

  // validate 统一为 Validator[]
  // 1、当 Validator 没有设置 validateTrigger，会使用 Field 的 validateTrigger
  // 2、在 1 条件下，当 field 没有设置validateTrigger，会使用默认的 validateTrigger
  const mergedValidators = (() => {
    const validators = (0, _shared.asArray)(validate).map(_shared.normalizeValidator);
    validators.forEach(v => {
      if (!v.validateTrigger.length) {
        v.validateTrigger = (0, _shared.asArray)(validateTrigger);
      }
      if (!v.validateTrigger.length) {
        v.validateTrigger.push(_shared.DEFAULT_VALIDATE_TRIGGER);
      }
    });
    return validators;
  })();
  const mergedValidatorsRef = (0, _hooks.useLatest)(mergedValidators);
  const allValidateTriggers = (() => {
    const triggerSet = new Set();
    mergedValidators.forEach(v => {
      v.validateTrigger.forEach(t => {
        if (t !== _shared.DEFAULT_VALIDATE_TRIGGER) {
          triggerSet.add(t);
        }
      });
    });
    return [...triggerSet];
  })();
  React.useEffect(() => {
    _registerField({
      fieldID,
      serializedName: namePathStr,
      validate: mergedValidatorsRef
    });
    return () => {
      _unregisterField(fieldID);
    };
  }, [_registerField, _unregisterField, mergedValidatorsRef, fieldID, namePathStr]);

  /**
   * Super mega handle Change
   */
  const handleChangeMega = React.useCallback(eventOrValue => {
    /**
     * 此处 forceUpdate 是因为用了 use-context-selector 导致 change 之后 Field 没有重新渲染，
     * 导致 child 拿到旧值，引发光标位置的抖动
     * issue： https://code.byted.org/universe-ui/react/issues/937
     */
    forceUpdate();
    // First arg is event
    let _value;
    if (typeof eventOrValue === 'object' && 'nativeEvent' in eventOrValue && eventOrValue.nativeEvent instanceof Event) {
      if (eventOrValue.target instanceof HTMLInputElement) {
        switch (eventOrValue.target.type) {
          case 'radio':
          case 'checkbox':
            _value = eventOrValue.target[valuePropName !== null && valuePropName !== void 0 ? valuePropName : 'checked'];
            break;
          case 'file':
            throw new Error('Field cannot be used on file inputs since they are read-only. Use a onChange listener instead.');
          default:
            _value = eventOrValue.target[valuePropName !== null && valuePropName !== void 0 ? valuePropName : 'value'];
        }
      } else if (eventOrValue.target instanceof HTMLSelectElement) {
        if (!eventOrValue.target.multiple) {
          _value = eventOrValue.target.value;
        } else {
          const res = [];
          for (let i = 0; i < eventOrValue.target.options.length; i++) {
            const opt = eventOrValue.target.options[i];
            if (opt.selected) {
              res.push(opt.value);
            }
          }
          _value = res;
        }
      } else {
        // radio group
        switch (eventOrValue.target.type) {
          case 'radio':
          case 'checkbox':
            _value = eventOrValue.target[valuePropName !== null && valuePropName !== void 0 ? valuePropName : 'checked'];
            break;
          default:
            _value = eventOrValue.target[valuePropName !== null && valuePropName !== void 0 ? valuePropName : 'value'];
            break;
        }
      }
    } else {
      _value = eventOrValue;
    }
    _setFieldTouched(namePathStr, true);
    _setFieldValue(namePathStr, _value, {
      triggerByUser: true,
      triggerName: trigger,
      shouldValidate: true
    });
  }, [_setFieldTouched, _setFieldValue, forceUpdate, namePathStr, trigger, valuePropName]);
  const validateField = React.useCallback(options => {
    validate$.next({
      serializedName: namePathStr,
      triggerName: (options === null || options === void 0 ? void 0 : options.triggerName) || _shared.DEFAULT_VALIDATE_TRIGGER
    });
  }, [namePathStr, validate$]);
  return {
    value,
    handleChange: handleChangeMega,
    error,
    allValidateTriggers,
    serializedName: namePathStr,
    trigger,
    validateField
  };
}
