// @universe-design/react/src/transfer/panel/SourceList.tsx
import List from '@universe-design/rc-virtual-list';
import * as React from 'react';
import clsx from 'clsx';
import stylex from '@universe-design/stylex';
import { useTransferContext } from '../hooks/TransferContext';
import { flatTransferGroups, TransferListType } from '../utils';
import { useTransferPanelContext } from '../hooks/PanelContext';
import { TransferListCaption } from './TransferListCaption';
import { sxPanelListWrapper, sxPanelEmpty, sxPanelListAutoSizer, sxPanelListScrolling } from '../sx';
import { TransferSourceOption } from './TransferSourceOption';
import { TransferOptionProvider } from './OptionProvider';
import { AutoSizer } from '../../shared/components/AutoSizer';
import { HighlightProvider } from '../../shared/components/highlighter';
import { useScrollStatus } from '../../hooks/useScrollStatus';
import { composeEventHandlers } from '../../util/function';
import { useRcVirtualListResizeObserver } from '../../hooks/useRcVirtualListResizeObserver';
import { forwardHTMLRef } from '../../util/forwardRef';
import type { TransferListProps } from '../interface';

/**
 * 默认AutoSizeDefaultHeight为0，数据量大的时候会导致rc-virtual-list 计算所有item的高度，造成浏览器假死
 * 设置一个默认10行高度的值，渲染过程中 AutoSize 会重新计算正确的高度
 */
const autoSizeDefaultHeight = 32 * 10;

export const TransferSourceList = forwardHTMLRef<TransferListProps, 'div'>((props, ref) => {
  /** Context */
  const { cssScope, filteredOptions, groupWeakMap, group, searchValue } = useTransferContext();
  const { emptyContent } = useTransferPanelContext();

  /** Props */
  const { className, listRef, ...htmlProps } = props;

  /** Derived value */
  const showGroup = Array.isArray(group) ? group[0] : group;
  const flatFilterOptions = React.useMemo(
    () => flatTransferGroups(filteredOptions, groupWeakMap, showGroup),
    [filteredOptions, groupWeakMap, showGroup],
  );

  /** Scroll status */
  const autoSizerRef = React.useRef<HTMLDivElement>(null);
  const { overflow, checkOverflow } = useScrollStatus();
  const { vListHolderRef } = useRcVirtualListResizeObserver({
    getContainer: () => autoSizerRef.current,
    onHolderResize: () => {
      checkOverflow(vListHolderRef.current);
    },
    onHolderChildResize: () => {
      checkOverflow(vListHolderRef.current);
    },
  });

  /** List node */
  let listNode;
  if (flatFilterOptions.length === 0) {
    /** Empty */
    listNode = (
      <div {...htmlProps} className={clsx(cssScope('__panel__empty'), className, stylex(sxPanelEmpty))}>
        {emptyContent}
      </div>
    );
  } else {
    /** VList */
    listNode = (
      <AutoSizer
        className={stylex(sxPanelListAutoSizer)}
        ref={autoSizerRef}
        defaultHeight={autoSizeDefaultHeight}
        disableWidth
      >
        {({ height }) => (
          <List
            className={clsx(
              cssScope('__panel__list'),
              overflow.t && [cssScope('__panel__list--scrolling'), stylex(sxPanelListScrolling)],
            )}
            itemKey={item =>
              item.type === TransferListType.CAPION ? `group_key_${item.group.key}` : item.option.value
            }
            data={flatFilterOptions}
            ref={listRef}
            virtual
            height={height}
            fullHeight
            itemHeight={32}
            onScroll={composeEventHandlers(props.onScroll, ev => {
              checkOverflow(ev.currentTarget);
            })}
          >
            {item => {
              if (item.type === TransferListType.CAPION) {
                return <TransferListCaption>{item.group.title}</TransferListCaption>;
              }
              return (
                // item 必须有 ref
                <div>
                  <TransferOptionProvider option={item.option}>
                    <TransferSourceOption />
                  </TransferOptionProvider>
                </div>
              );
            }}
          </List>
        )}
      </AutoSizer>
    );
  }

  /** List */
  return (
    <HighlightProvider
      value={{
        searchWords: [searchValue],
        disabled: false,
      }}
    >
      <div
        {...htmlProps}
        className={clsx(cssScope('__panel__list__wrapper'), className, stylex(sxPanelListWrapper))}
        ref={ref}
      >
        {listNode}
      </div>
    </HighlightProvider>
  );
});
