// @universe-design/react/src/transfer/panel/SourcePanel.tsx
import stylex from '@universe-design/stylex';
import * as React from 'react';
import clsx from 'clsx';
import { useTransferContext } from '../hooks/TransferContext';
import { TransferSearch } from './Search';
import { Checkbox } from '../../checkbox';
import { TransferPanelHeader, PanelContainer } from './PanelContainer';
import { TransferSourceList } from './SourceList';
import type { TransferOption, TransferPanelProps, TransferValue } from '../interface';
import { ListItemLike } from '../../shared/components/list';
import { useLocaleReceiver } from '../../config-provider';
import { forwardHTMLRef } from '../../util/forwardRef';
import { sxSelectAll } from '../sx';

/**
 * @public
 * `TransferSelectAll` 的类型。
 */
export interface TransferSelectAllProps {
  /**
   * 选择全部文案
   * @defaultValue `All (${selected}/${total})`
   */
  label?: (selectedCount: number, totalCount: number) => React.ReactNode;
}

export const TransferSelectAll = forwardHTMLRef<TransferSelectAllProps, 'div'>((props, ref) => {
  /** Context */
  const { cssScope, toggleSelectAll, isAllSelected, value, filteredOptions } = useTransferContext();

  /** Locale */
  const { selectAll } = useLocaleReceiver('transfer');

  if (filteredOptions.length === 0) {
    return null;
  }

  /** Props */
  const {
    label = (selected: number, total: number) => selectAll.f({ selected: String(selected), total: String(total) }),
    className,
    ...htmlProps
  } = props;

  /** Checked status */
  let indeterminate = false;
  const checked = isAllSelected;
  if (!checked && value.length) {
    indeterminate = filteredOptions.some(op => value.includes(op.value));
  }

  return (
    <div ref={ref} className={clsx(className, cssScope('__select-all'), stylex(sxSelectAll))} {...htmlProps}>
      <ListItemLike>
        <Checkbox
          indeterminate={indeterminate}
          checked={checked}
          onCheckChange={toggleSelectAll}
          data-testid="transfer-select-all"
        >
          {/* TODO 为什么用 filteredOptions 做总数？ */}
          {label(value.length, filteredOptions.length)}
        </Checkbox>
      </ListItemLike>
    </div>
  );
});

export const TransferSourcePanelHeader = forwardHTMLRef<unknown, 'div'>((props, ref) => {
  /** Memo default children */
  const defaultChildren = React.useMemo(
    () => (
      <>
        <TransferSearch />
        <TransferSelectAll />
      </>
    ),
    [],
  );

  /** Props */
  const { children = defaultChildren, ...htmlProps } = props;

  return (
    <TransferPanelHeader {...htmlProps} ref={ref}>
      {children}
    </TransferPanelHeader>
  );
});

export const TransferSourcePanel = forwardHTMLRef<TransferPanelProps, 'div'>((props, ref) => {
  /** Memo default children */
  const defaultChildren = React.useMemo(
    () => (
      <>
        <TransferSourcePanelHeader />
        <TransferSourceList />
      </>
    ),
    [],
  );

  /** Props */
  const { children = defaultChildren, optionRenderer, emptyContent: noOptionContentProp, ...htmlProps } = props;

  /** Context */
  const { searchValue, cssScope } = useTransferContext();

  /** Locale */
  const { noSearchResult, noOptions } = useLocaleReceiver('transfer');
  const emptyContent = noOptionContentProp ?? (searchValue ? noSearchResult : noOptions);

  return (
    <PanelContainer
      {...htmlProps}
      emptyContent={emptyContent}
      optionRenderer={optionRenderer}
      ref={ref}
      className={clsx(cssScope('__source-panel'), props.className)}
    >
      {children}
    </PanelContainer>
  );
}) as <V extends TransferValue, Op extends TransferOption<V>>(
  props: React.PropsWithChildren<TransferPanelProps<V, Op>> & {
    ref?: React.Ref<HTMLDivElement>;
  },
) => React.ReactElement;
