// @universe-design/react/src/transfer/panel/TargetPanel.tsx
import stylex from '@universe-design/stylex';
import * as React from 'react';
import clsx from 'clsx';
import { Button } from '../../button';
import { useTransferContext } from '../hooks/TransferContext';
import { TransferPanelHeader, PanelContainer } from './PanelContainer';
import { TransferTargetList } from './TargetList';
import type { TransferOption, TransferPanelProps, TransferValue } from '../interface';
import { ListItemLike } from '../../shared/components/list';
import { useLocaleReceiver } from '../../config-provider';
import { forwardHTMLRef } from '../../util/forwardRef';
import { sxClearButton, sxSelectionSummary } from '../sx';

export interface TranferSelectionSummaryProps {
  /**
   * 是否显示清除按钮
   * @defaultValue true
   */
  showClearButton?: boolean;
  /**
   * 自定义已选数量统计文案
   * @defaultValue `Selected (${total})`
   */
  selectionSummary?: (selectedCount: number) => React.ReactNode;
}

export const TranferSelectionSummary = forwardHTMLRef<TranferSelectionSummaryProps, 'div'>((props, ref) => {
  /** Locale */
  const { selected, clear } = useLocaleReceiver('transfer');

  /** Props */
  const {
    showClearButton = true,
    selectionSummary = (selectedCount: number) => selected.f({ selected: String(selectedCount) }),
    ...htmlProps
  } = props;

  /** Context */
  const { cssScope, value, clearAll } = useTransferContext();

  return (
    <div className={clsx(stylex(sxSelectionSummary), cssScope('__selection-summary'))} {...htmlProps} ref={ref}>
      <ListItemLike>{selectionSummary(value.length)}</ListItemLike>
      {showClearButton && value.length > 0 && (
        <Button
          type="text"
          onClick={clearAll}
          color="primary"
          data-testid="transfer-clear-all"
          className={stylex(sxClearButton)}
        >
          {clear}
        </Button>
      )}
    </div>
  );
});

export const TransferTargetPanelHeader = forwardHTMLRef<unknown, 'div'>((props, ref) => {
  /** Memo default children */
  const defaultChildren = React.useMemo(() => <TranferSelectionSummary />, []);

  /** Props */
  const { children = defaultChildren, ...htmlProps } = props;

  return (
    <TransferPanelHeader {...htmlProps} ref={ref}>
      {children}
    </TransferPanelHeader>
  );
});

export const TransferTargetPanel = forwardHTMLRef<TransferPanelProps, 'div'>((props, ref) => {
  /** Memo default children */
  const defaultChildren = React.useMemo(
    () => (
      <>
        <TransferTargetPanelHeader />
        <TransferTargetList />
      </>
    ),
    [],
  );

  /** Props */
  const { children = defaultChildren, optionRenderer, emptyContent: noOptionContentProp, ...htmlProps } = props;

  /** Context */
  const { cssScope } = useTransferContext();

  /** Locale */
  const { noOptionsSelected } = useLocaleReceiver('transfer');

  return (
    <PanelContainer
      {...htmlProps}
      emptyContent={noOptionsSelected ?? noOptionContentProp}
      optionRenderer={optionRenderer}
      ref={ref}
      className={clsx(cssScope('__target-panel'), props.className)}
    >
      {children}
    </PanelContainer>
  );
}) as <V extends TransferValue = string, OP extends TransferOption<V> = TransferOption<V>>(
  props: React.PropsWithChildren<TransferPanelProps<V, OP>> & {
    ref?: React.Ref<HTMLDivElement>;
  },
) => React.ReactElement;
