// @universe-design/styles/esm/lark-minutes/index.light.atom.js
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
export const Minutestoken_bg_recording = '#fcfcfc';
export const Minutestoken_home_card_Iceblue_bg = '#99acc7';
export const Minutestoken_home_card_Iceblue_highlight_bg = '#a2e9ff';
export const Minutestoken_home_card_fogblue_bg = '#dce5fa';
export const Minutestoken_home_card_gold_bg = '#f7f1e1';
export const Minutestoken_home_card_gold_highlight_bg = '#fcdca4';
export const Minutestoken_home_card_gray_bg = '#f0f1f2';
export const Minutestoken_home_card_gray_highlight_bg = '#ff9d4c';
export const Minutestoken_home_card_green_bg = '#e3f1f0';
export const Minutestoken_home_card_green_highlight_bg = '#efe6fe';
export const Minutestoken_home_card_indigo_bg = '#99a3c4';
export const Minutestoken_home_card_indigo_highlight_bg = '#86d1ff';
export const Minutestoken_home_card_skyblue_bg = '#e6eef7';
export const Minutestoken_home_card_skyblue_highlight_bg = '#c2d4ff';
export const Minutestoken_home_card_text_on_dark = '#ffffff';
export const Minutestoken_home_card_text_on_light = 'rgba(31, 35, 41, 0.8)';
export const Minutestoken_meeting_onboarding_text_title = '#333dcc';
export const Minutestoken_notes_bg_base = '#f2f3f5';
export const Minutestoken_notes_bg_device = 'linear-gradient(to top, rgba(49, 50, 89, 0.6) 0%, rgba(242, 243, 245, 0) 100%)';
export const Minutestoken_notes_bg_gradient_01 = 'linear-gradient(45deg, #e9eafb 0%, #f0f4ff 100%)';
export const Minutestoken_notes_bg_gradient_02 = 'linear-gradient(45deg, #cccff9 0%, #e0e9ff 100%)';
export const Minutestoken_notes_content_gradient = 'linear-gradient(45deg, #6364f9 0%, #4c81fc 100%)';
export const Minutestoken_notes_fill_solid_01 = '#f2f3fd';
export const Minutestoken_notes_fill_solid_02 = '#e9eafb';
export const Minutestoken_notes_fill_transparent_01 = 'rgba(91, 101, 245, 0.1)';
export const Minutestoken_notes_fill_transparent_02 = 'rgba(91, 101, 245, 0.15)';
export const Minutestoken_notes_primary_content_default = '#4752e6';
export const Minutestoken_notes_primary_content_hover = '#5b65f5';
export const Minutestoken_notes_primary_content_loading = '#abb0f2';
export const Minutestoken_notes_primary_content_pressed = '#333dcc';
export const Minutestoken_notes_primary_fill_default = '#5b65f5';
export const Minutestoken_notes_primary_fill_hover = '#757df0';
export const Minutestoken_notes_primary_fill_loading = '#abb0f2';
export const Minutestoken_notes_primary_fill_pressed = '#4752e6';
export const Minutestoken_notes_recording_btn_bg = 'linear-gradient(to right, #555ff2 0%, #8b76f5 100%)';
export const Minutestoken_notes_recording_btn_hover = 'linear-gradient(to right, #6871f2 0%, #a08ef5 100%)';
export const Minutestoken_notes_recording_btn_loading = 'linear-gradient(to right, #939df5 0%, #ae9ff5 100%)';
export const Minutestoken_notes_recording_btn_pressed = 'linear-gradient(to right, #4652eb 0%, #826ded 100%)';
export const lmtoken_float_bg_body = 'rgba(255, 255, 255, 0.6)';
export const lmtoken_float_btn_border_blue = '#c2d4ff';
export const lmtoken_recording_btn_border_gray = 'rgba(31, 35, 41, 0.3)';
export const lmtoken_recording_btn_small_gray = 'rgba(31, 35, 41, 0.05)';
