// @universe-design/react/src/message/index.tsx
import React, { useRef } from 'react';
import ReactDOM from 'react-dom';
import { CrossRootConfigProvider, ConfigProvider } from '../config-provider';
import type {
  ToastInstance,
  ToastProps,
  ToastContentConfig,
  ConfigDuration,
  ToastConfigOrContentType,
  ToastMethodConfig,
} from './Toast';
import { Toast } from './Toast';

import './style';

let globalToast: Toast | undefined;
let globalToastInitPromise: Promise<Toast> | undefined;

function getDefaultContainer() {
  const div = document.createElement('div');
  document.body.appendChild(div);
  return div;
}

function singletonRef(elem: Toast | null) {
  if (elem) {
    globalToast = elem;
  }
}

/**
 * 用于多次重复 renderGlobalToast 和销毁自动生成的 DOM 元素
 */
class ToastWrapper extends React.Component<ToastProps> {
  private div?: HTMLElement;
  private isDefaultContainer = true;

  componentWillUnmount() {
    globalToast = undefined;
    if (this.isDefaultContainer && this.div) {
      this.div.remove();
      this.div = undefined;
    }
  }

  getContainer = () => {
    if (!this.div) {
      const globalContainer = Toast.globalConfig.getContainer?.();
      this.div = globalContainer || getDefaultContainer();
      if (globalContainer) {
        this.isDefaultContainer = false;
      }
    }
    return this.div;
  };

  render() {
    return <Toast {...this.props} ref={singletonRef} getContainer={this.getContainer} />;
  }
}

export function getGlobalToast(): Promise<Toast> {
  if (globalToast) {
    return Promise.resolve(globalToast);
  }
  if (globalToastInitPromise) {
    return globalToastInitPromise;
  }
  globalToastInitPromise = new Promise(resolve => {
    const div = getDefaultContainer();
    // 如果没有使用 useToast 或者 renderGlobalToast 初始化 toast，
    // 则默认取第一个 ConfigProvider 提供的配置
    ReactDOM.render(
      <CrossRootConfigProvider>
        <ToastWrapper
          ref={() => {
            if (globalToast) {
              globalToastInitPromise = undefined;
              resolve(globalToast);
            }
          }}
          getContainer={() => div}
        />
      </CrossRootConfigProvider>,
      div,
    );
  });
  return globalToastInitPromise;
}

function createTostWrapper(getToast: () => Promise<Toast>) {
  return {
    open(config: ToastContentConfig) {
      return getToast().then(toast => toast.open(config) || '');
    },

    info(configOrContent: ToastConfigOrContentType, duration?: ConfigDuration, onClose?: () => void) {
      return getToast().then(toast => toast.info(configOrContent, duration, onClose));
    },

    success(configOrContent: ToastConfigOrContentType, duration?: ConfigDuration, onClose?: () => void) {
      return getToast().then(toast => toast.success(configOrContent, duration, onClose));
    },

    warning(configOrContent: ToastConfigOrContentType, duration?: ConfigDuration, onClose?: () => void) {
      return getToast().then(toast => toast.warning(configOrContent, duration, onClose));
    },

    warn(configOrContent: ToastConfigOrContentType, duration?: ConfigDuration, onClose?: () => void) {
      return getToast().then(toast => toast.warn(configOrContent, duration, onClose));
    },

    error(configOrContent: ToastConfigOrContentType, duration?: ConfigDuration, onClose?: () => void) {
      return getToast().then(toast => toast.error(configOrContent, duration, onClose));
    },

    loading(configOrContent: ToastConfigOrContentType, duration?: ConfigDuration, onClose?: () => void) {
      return getToast().then(toast => toast.loading(configOrContent, duration, onClose));
    },

    config(options: ToastProps): void {
      Toast.globalConfig = Object.assign(Toast.globalConfig, options);
    },

    destroy(key?: React.Key): void {
      getToast().then(toast => toast.destroy(key));
    },
  };
}

export { Toast };

export const toast = createTostWrapper(getGlobalToast);

/**
 * 用于创建可以获取 ConfigContext 的全局 Toast 单例
 * @example
 * ```tsx
 * <ConfigProvider>
 *   {renderGlobalToast({ className: 'my-toast' }, document.getElementById('my-container'))}
 *   <App />
 * </ConfigProvider>
 * ```
 *
 * @param props: ToastProps
 * @returns React.ReactPortal
 */
export function renderGlobalToast(props?: ToastProps) {
  return <ToastWrapper {...props} />;
}

/**
 * 获取一个局部的 toast 实例
 * @returns
 */
export function useToast(props?: ToastProps, container?: HTMLElement): [ToastInstance, React.ReactNode] {
  const toastRef = useRef<Toast>(null);
  const mountDivRef = useRef<HTMLDivElement>();
  const toastIns = createTostWrapper(() => {
    if (toastRef.current) {
      return Promise.resolve(toastRef.current);
    }
    return getGlobalToast();
  });
  const _getDefaultContainer = () => {
    if (mountDivRef.current) {
      return mountDivRef.current;
    }
    mountDivRef.current = getDefaultContainer();
    return mountDivRef.current;
  };
  return [
    toastIns,
    <ConfigProvider key="ud-hook-toast-wrapper">
      <Toast key="ud-hook-toast" {...props} ref={toastRef} getContainer={() => container || _getDefaultContainer()} />
    </ConfigProvider>,
  ];
}

// 兼容竞品 API
export const message = toast;
export const useMessage = useToast;

export type {
  ToastProps,
  ToastContentConfig,
  ToastMethodConfig,
  ConfigDuration,
  ToastConfigOrContentType,
  ToastInstance,
};
