// @universe-design/react/src/message/Toast.tsx
import type { ReactChild } from 'react';
import React, { createRef } from 'react';
import clsx from 'clsx';
import omit from 'lodash/omit';
import pick from 'lodash/pick';
import type { CSSTransitionProps } from 'react-transition-group/CSSTransition';
import stylex from '@universe-design/stylex';
import { cssTypographyHeadline } from '../stylex';
import { Notice } from '../notice';
import type { PresetStatusColorType } from '../util';
import type { MessageConfig } from '../notification/manager';
import type { Slot } from '../shared/slot';
import type { ButtonProps } from '../button';
import NotificationManager, {
  // BASE_MESSAGE_LENGTH,
  isNotAutoClose,
} from '../notification/manager';

import './style';

export type ToastConfigOrContentType = ReactChild | null | undefined | ToastMethodConfig;

const ManagerProps = [
  'className',
  'prefixCls',
  'duration',
  'transitionProps',
  'maxCount',
  'getContainer',
  'top',
] as const;

/**
 * `toast.config` 接收的参数。
 * @public
 * @docgen
 */
export interface ToastProps {
  /**
   * 是否可关闭
   * @defaultValue false
   */
  // closable?: boolean;
  // /**
  //  * 是否显示图标，这里和其他开源框架不同，UD 默认为 true。并且 UD 规范不推荐使用
  //  * @defaultValue true
  //  * @deprecated
  //  */
  // showIcon?: boolean;
  // /**
  //  * `style` 属性透传。
  //  */
  // style?: React.CSSProperties;
  /**
   *  `className` 属性透传。
   */
  className?: string;
  /**
   * 替换默认的 class prefix
   */
  prefixCls?: string;
  /**
   * 持续的时长，单位是毫秒，配置为 0 或 null 则不自动关闭。
   * @defaultValue 4000, 当传入 action 时为 6000
   */
  duration?: number | null;
  /**
   * CSSTransition 的 props
   */
  transitionProps?: CSSTransitionProps;
  /**
   * 最多显示多少个消息，超过的消息将在队列中维护
   */
  maxCount?: number;
  /**
   * 距离顶部距离
   */
  top?: number;
  /**
   * 包含 toast 和 notification 的容器
   */
  getContainer?: () => HTMLElement;
  /**
   * RTL 模式，暂时不支持
   * @deprecated
   */
  rtl?: boolean;
}
/**
 * toast 内容配置。
 * @public
 * @docgen
 */
export interface ToastContentConfig extends Pick<ToastProps, 'duration'> {
  key?: React.Key;
  /**
   *  是否显示关闭按钮，默认 `false`
   */
  showCloseIcon?: boolean;
  /**
   * `style` 属性透传。
   */
  style?: React.CSSProperties;
  /**
   *  `className` 属性透传。
   */
  className?: string;
  /**
   * 状态类型
   * @defaultValue `info`
   */
  type?: PresetStatusColorType | 'loading';
  /**
   * 跟在提示内容后面的操作按钮
   */
  action?: React.ReactNode;
  /**
   * 内容
   */
  content: React.ReactNode;
  /**
   * `mouseenter` 事件。
   */
  onMouseEnter?: React.MouseEventHandler<HTMLDivElement>;
  /**
   * `mouseleave` 事件。
   */
  onMouseLeave?: React.MouseEventHandler<HTMLDivElement>;
  /**
   * `click` 事件。
   */
  onClick?: React.MouseEventHandler<HTMLDivElement>;
  onClose?: () => void;
  /**
   * 替换状态按钮。
   */
  icon?: React.ReactNode;
  /**
   * @deprecated 用于风格切换过渡期同时提供两种样式，即将废弃，请使用默认样式
   */
  newStyle?: boolean;
  /**
   * 关闭按钮
   */
  closeButton?: Slot<ButtonProps>;
}

export type ToastMethodConfig = Omit<ToastContentConfig, 'type'>;

export type ConfigDuration = ToastContentConfig['duration'] | (() => ToastContentConfig['duration']);

export interface ToastInstance {
  // open a toast, return a promise resolved after close
  open: (config: ToastContentConfig) => Promise<void>;
  info: (configOrContent: ToastConfigOrContentType, duration?: ConfigDuration, onClose?: () => void) => Promise<void>;
  success: (
    configOrContent: ToastConfigOrContentType,
    duration?: ConfigDuration,
    onClose?: () => void,
  ) => Promise<void>;
  error: (configOrContent: ToastConfigOrContentType, duration?: ConfigDuration, onClose?: () => void) => Promise<void>;
  warning: (
    configOrContent: ToastConfigOrContentType,
    duration?: ConfigDuration,
    onClose?: () => void,
  ) => Promise<void>;
  warn: (configOrContent: ToastConfigOrContentType, duration?: ConfigDuration, onClose?: () => void) => Promise<void>;
  loading: (
    configOrContent: ToastConfigOrContentType,
    duration?: ConfigDuration,
    onClose?: () => void,
  ) => Promise<void>;
  config: (props: ToastProps) => void;
  destroy: (key?: React.Key) => void;
}

function getActionDuration(action: any, duration?: number | null) {
  // eslint-disable-next-line @typescript-eslint/no-magic-numbers
  const actionDur = action ? 6000 : undefined;
  return duration !== undefined ? duration : actionDur;
}

function getManagerConfig(option: ToastContentConfig): MessageConfig {
  const { content, key, duration, onClose, className, type, showCloseIcon, newStyle = true, ...rest } = option;
  // let contentKey: string | undefined;
  // let contentLength = 0;
  // if (typeof content === 'string') {
  //   contentKey = `${type}-${content}`;
  //   contentLength = content.length;
  // }
  // // 显示关闭按钮的逻辑
  // // 1. 设置不自动关闭时
  // // 2. 全局配置 closable 为 true
  // // 3. 内容长度 >= BASE_MESSAGE_LENGTH
  // // 4. 内容是 object 类型时，包含 null
  // const newClosable =
  //   isNotAutoClose(duration) ||
  //   showCloseIcon ||
  //   contentLength >= BASE_MESSAGE_LENGTH ||
  //   typeof content === 'object';
  const newClosable = showCloseIcon !== undefined ? showCloseIcon : isNotAutoClose(duration);
  return {
    key,
    duration: getActionDuration(rest.action, duration),
    renderMessage: ({ close, rootPrefixCls, toastConfigCls }) => (
      <Notice
        {...rest}
        closable={newClosable}
        visible
        className={clsx(`${rootPrefixCls}__toast`, className, toastConfigCls || '')}
        theme={newStyle ? 'outlined' : 'plain'}
        message={<div className={stylex(cssTypographyHeadline)}>{content}</div>}
        type={type}
        onClose={close}
        banner={false}
        // maxLine={2}
      />
    ),
    onClose,
  };
}

/**
 * 1. 不需要 theme 参数，只有一种样式
 * 2. info 使用目前的 normal 样式
 * 3. 其它使用目前的 light 模式
 * 4. export Toast 可挂载到当前容器
 * 5. export message、toast 直接调起 API
 */
export class Toast extends React.PureComponent<ToastProps> implements ToastInstance {
  managerRef = createRef<NotificationManager>();

  /**
   * 全局 config
   */
  static globalConfig: ToastProps = {};

  /**
   * 计算后的配置
   */
  private get _config() {
    return Object.assign({}, Toast.globalConfig, this.props);
  }

  /* eslint-disable no-dupe-class-members */
  /**
   * 返回一个 onClose 执行时 resolve 的 promise
   * @param config
   * @returns
   */
  open(config: ToastContentConfig): Promise<void> {
    return new Promise(resolve => {
      const handleClose = config.onClose;
      config.onClose = () => {
        handleClose && handleClose();
        resolve(undefined);
      };
      this.managerRef.current?.add(this._getManagerConfig(config));
    });
  }

  info(configOrContent: ToastConfigOrContentType, duration?: ConfigDuration, onClose?: () => void): Promise<void> {
    return this._open('info', configOrContent, duration, onClose);
  }

  success(configOrContent: ToastConfigOrContentType, duration?: ConfigDuration, onClose?: () => void): Promise<void> {
    return this._open('success', configOrContent, duration, onClose);
  }

  warning(configOrContent: ToastConfigOrContentType, duration?: ConfigDuration, onClose?: () => void): Promise<void> {
    return this._open('warning', configOrContent, duration, onClose);
  }

  warn(configOrContent: ToastConfigOrContentType, duration?: ConfigDuration, onClose?: () => void): Promise<void> {
    return this.warning(configOrContent, duration, onClose);
  }

  error(configOrContent: ToastConfigOrContentType, duration?: ConfigDuration, onClose?: () => void): Promise<void> {
    return this._open('error', configOrContent, duration, onClose);
  }

  loading(configOrContent: ToastConfigOrContentType, duration?: ConfigDuration, onClose?: () => void): Promise<void> {
    return this._open('loading', configOrContent, duration, onClose);
  }

  config(props: ToastProps): void {
    // TODO 这里没有想好，之前是 assign 但是感觉直接覆盖 config 比较好，要不有些值写入了就没法删掉了
    Toast.globalConfig = Object.assign(Toast.globalConfig, props);
    // Toast.config = props;
    this.forceUpdate();
  }

  destroy(key?: React.Key): void {
    if (key) {
      this.managerRef.current?.remove(key);
    } else {
      this.managerRef.current?.destroy();
    }
  }

  render() {
    // Debug
    // console.log('Toast render', { config: this._config });
    return (
      <NotificationManager
        ref={this.managerRef}
        placement="top"
        transitionProps={{
          timeout: 300,
          classNames: 'toast--animation',
        }}
        duration={3000}
        {...pick(this._config, ManagerProps)}
      />
    );
  }

  private _getManagerConfig(config: ToastContentConfig) {
    const props = Object.assign({}, omit(this._config, ManagerProps));
    return getManagerConfig({
      ...props,
      ...config,
    });
  }

  private _open(
    type: ToastContentConfig['type'],
    configOrContent: React.ReactNode | ToastMethodConfig,
    duration?: ConfigDuration,
    onClose?: () => void,
  ): Promise<void> {
    if (
      configOrContent &&
      typeof configOrContent === 'object' &&
      (('content' in configOrContent) as unknown as ToastContentConfig)
    ) {
      return this.open({ ...(configOrContent as ToastContentConfig), type });
    } else {
      const d = typeof duration === 'function' ? duration() : duration;
      return this.open({
        type,
        content: configOrContent,
        duration: d,
        onClose,
      });
    }
  }
  /* eslint-enable no-dupe-class-members */
}
