// @universe-design/react/es/date-picker/panels/DateCalendar/DateCalendarUI.js
import _extends from "@babel/runtime/helpers/esm/extends";
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx } from '@universe-design/stylex/react';
import stylex from '@universe-design/stylex';
import clsx from 'clsx';
import isFunction from 'lodash/isFunction';
import isNil from 'lodash/isNil';
import min from 'lodash/min';
import noop from 'lodash/noop';
import moment from 'moment';
import React, { useEffect } from 'react';
import { useControlledState } from '@universe-design/react-utilities';
import { useCssScope } from '../../../config-provider';
import { Space } from '../../../space';
import { forwardHTMLRef } from '../../../util/forwardRef';
import { sxPickerHeaderSwitchWrapper } from '../../sx';
import { Next20YearsSwitcher, NextMonthSwitcher, NextYearSwitcher, Pre20YearsSwitcher, PreMonthSwitcher, PreYearSwitcher } from '../PanelHeaderSwitchers';
import { getHigherLevelMode, getNext, getPrev } from '../base/utils';
import Body from '../picker/Body';
import { MonthYearSelector } from '../selector/MonthYearSelector';
import { CalendarUIProvider } from './DateCalendarUIContext';
import '../../style';
import { shouldCellDisabled as shouldCellDisabledFn } from '../../utils/disabled';
export const isDateValueEqual = (a, b) => {
  if (isNil(a) && isNil(b)) {
    return true;
  }
  if (a === b || a !== null && a !== void 0 && a.isSame(b)) {
    return true;
  }
  return false;
};
const defaultPreviewDate = () => false;
const defaultDisabledDate = () => false;
const DateCalendarUI = forwardHTMLRef((props, ref) => {
  const {
    className,
    style,
    mode = 'date',
    pickerValue: pickerValueProp,
    hiddenCollapse,
    // monthYearSelectorProps
    onCollapsed,
    onMonthClick,
    onYearClick,
    // bodyProps
    onPickerValueChange,
    hiddenNext,
    hiddenPrev,
    headerRenderer,
    // context start
    defaultPickerValue,
    // pickerValue,
    selectedDates = [],
    selectedRange,
    onDateMouseEnter = noop,
    onDateMouseLeave = noop,
    onSelect = noop,
    previewDate = defaultPreviewDate,
    disabledDate = defaultDisabledDate,
    minDate,
    maxDate,
    startOfWeek = moment.localeData().firstDayOfWeek(),
    noOutOfViewWeek = false,
    cellRenderer,
    onContentMouseEnter = noop,
    onContentMouseLeave = noop,
    sx,
    janX,
    hideWeekNumbers = true,
    // context end
    ...restBodyProps
  } = props;
  const prefixCls = useCssScope('picker')();
  const datePanelPrefixCls = `${prefixCls}-date-panel`;
  const [localMode, setLocalMode] = React.useState(mode);
  useEffect(() => {
    if (!janX) {
      return;
    }
    const currentLocale = moment.locale();
    moment.updateLocale(currentLocale, {
      week: {
        dow: startOfWeek,
        doy: 7 + startOfWeek - janX
      }
    });
  }, [startOfWeek, janX]);

  // 当前面板切换到哪个时间点，并不是已选中的值或未提交的值
  const firstSelected = min([...selectedDates, ...(selectedRange || [])]);
  const [pickerValue = moment(), setViewDateState] = useControlledState(firstSelected !== null && firstSelected !== void 0 ? firstSelected : defaultPickerValue, pickerValueProp, {
    isEqual: isDateValueEqual
  });
  const setViewDate = date => {
    onPickerValueChange === null || onPickerValueChange === void 0 ? void 0 : onPickerValueChange(date, {
      mode: localMode
    });
    setViewDateState(date);
  };
  const shouldCellDisabled = React.useCallback((date, _mode) => shouldCellDisabledFn(date, {
    minDate,
    maxDate,
    disabledDate,
    mode: _mode
  }), [disabledDate, maxDate, minDate]);
  let PrevSwitcher, NextSwitcher;
  switch (localMode) {
    case 'date':
      PrevSwitcher = PreMonthSwitcher;
      NextSwitcher = NextMonthSwitcher;
      break;
    case 'month':
      PrevSwitcher = PreYearSwitcher;
      NextSwitcher = NextYearSwitcher;
      break;
    default:
      PrevSwitcher = Pre20YearsSwitcher;
      NextSwitcher = Next20YearsSwitcher;
      break;
  }
  // #region 自定义 Header 渲染
  const monthYearSelectorNode = jsx(MonthYearSelector, {
    mode: localMode,
    pickerValue: pickerValue,
    className: `${prefixCls}-panel-header-content__left`,
    hiddenCollapse: hiddenCollapse || localMode === 'year',
    collapsed: localMode === mode,
    onCollapsed: collapsed => {
      if (isFunction(onCollapsed)) {
        onCollapsed(collapsed);
      }
      if (collapsed) {
        setLocalMode(mode);
      } else {
        const nextMode = getHigherLevelMode(mode);
        if (nextMode) {
          setLocalMode(nextMode);
        }
      }
    },
    onYearClick: () => {
      if (isFunction(onYearClick)) {
        onYearClick();
      }
      setLocalMode('year');
    },
    onMonthClick: () => {
      if (isFunction(onMonthClick)) {
        onMonthClick();
      }
      setLocalMode('month');
    }
  });
  const prevBtnNode = jsx(PrevSwitcher, {
    onClick: () => setViewDate(getPrev(pickerValue, localMode))
  });
  const nextBtnNode = jsx(NextSwitcher, {
    onClick: () => setViewDate(getNext(pickerValue, localMode))
  });
  const defaultNode = jsx("div", {
    className: prefixCls + '-panel-header'
  }, jsx("div", {
    className: `${prefixCls}-panel-header-inner`
  }, monthYearSelectorNode, jsx(Space, {
    size: 12,
    className: stylex(sxPickerHeaderSwitchWrapper)
  }, !hiddenPrev && prevBtnNode, !hiddenNext && nextBtnNode)));
  const pickerHeaderNode = isFunction(headerRenderer) ? headerRenderer({
    prevBtnNode,
    nextBtnNode,
    defaultNode
  }) : defaultNode;
  // #endregion

  return jsx(CalendarUIProvider, {
    value: {
      // defaultPickerValue,
      onPickerValueChange,
      selectedDates,
      selectedRange,
      onDateMouseEnter,
      onDateMouseLeave,
      onSelect: newDate => {
        setViewDate(newDate);
        if (mode !== localMode) {
          setLocalMode(mode);
          return;
        }
        onSelect === null || onSelect === void 0 ? void 0 : onSelect(newDate);
      },
      previewDate,
      shouldCellDisabled,
      pickerValue,
      startOfWeek,
      noOutOfViewWeek,
      cellRenderer,
      onContentMouseEnter,
      onContentMouseLeave,
      janX,
      hideWeekNumbers
    }
  }, jsx("div", {
    className: clsx(className, datePanelPrefixCls, `${prefixCls}-date-panel__mode_${localMode}`, !hideWeekNumbers && localMode === 'date' && `${datePanelPrefixCls}__withWeekNumbers`),
    style: style,
    ref: ref,
    sx: sx
  }, pickerHeaderNode, jsx(Body, _extends({}, restBodyProps, {
    mode: localMode
  }))));
});
export { DateCalendarUI };
