// @universe-design/react/src/pagination/internal/Options.tsx
import React from 'react';
import KEYCODE from './KeyCode';
import type { PaginationLocale } from './types';
import stylex from '@universe-design/stylex';

interface OptionsProps {
  rootPrefixCls: string;
  locale: PaginationLocale;
  changeSize?: (size: number) => unknown;
  goButton: React.ReactNode;
  quickGo?: (page: number | undefined) => unknown;
  selectComponentClass: React.ComponentType<any> & {
    Option: React.ComponentType<any>;
  };
  pageSize: number;
  pageSizeOptions: string[];
  disabled?: boolean;
  selectPrefixCls?: string;
}

class Options extends React.Component<OptionsProps, { goInputText: string }> {
  state = {
    goInputText: '',
  };

  getValidValue() {
    const { goInputText } = this.state;
    return !goInputText || isNaN(Number(goInputText)) ? undefined : Number(goInputText);
  }

  buildOptionText = (value: string) => `${value} ${this.props.locale.items_per_page}`;

  changeSize = (value: unknown) => {
    this.props.changeSize!(Number(value));
  };

  handleChange: React.ChangeEventHandler<HTMLInputElement> = e => {
    this.setState({
      goInputText: e.target.value,
    });
  };

  handleBlur: React.FocusEventHandler<HTMLInputElement> = e => {
    const { goButton, quickGo, rootPrefixCls } = this.props;
    const { goInputText } = this.state;
    if (goButton || goInputText === '') {
      return;
    }
    this.setState({
      goInputText: '',
    });
    if (
      e.relatedTarget &&
      (e.relatedTarget.className.indexOf(`${rootPrefixCls}-item-link`) >= 0 ||
        e.relatedTarget.className.indexOf(`${rootPrefixCls}-item`) >= 0)
    ) {
      return;
    }
    quickGo?.(this.getValidValue());
  };

  go = (e: React.KeyboardEvent | React.MouseEvent) => {
    const { goInputText } = this.state;
    if (goInputText === '') {
      return;
    }
    if (('keyCode' in e && e.keyCode === KEYCODE.ENTER) || e.type === 'click') {
      this.setState({
        goInputText: '',
      });
      this.props.quickGo?.(this.getValidValue());
    }
  };

  getPageSizeOptions() {
    const { pageSize, pageSizeOptions } = this.props;
    if (pageSizeOptions.some(option => option.toString() === pageSize.toString())) {
      return pageSizeOptions;
    }
    return pageSizeOptions.concat([pageSize.toString()]).sort((a, b) => {
      const numberA = isNaN(Number(a)) ? 0 : Number(a);
      const numberB = isNaN(Number(b)) ? 0 : Number(b);
      return numberA - numberB;
    });
  }

  render() {
    const {
      pageSize,
      locale,
      rootPrefixCls,
      changeSize,
      quickGo,
      goButton,
      selectComponentClass,
      selectPrefixCls,
      disabled,
    } = this.props;
    const { goInputText } = this.state;
    const prefixCls = `${rootPrefixCls}-options`;
    const Select = selectComponentClass;
    let changeSelect = null;
    let goInput = null;
    let gotoButton = null;

    if (!changeSize && !quickGo) {
      return null;
    }

    const pageSizeOptions = this.getPageSizeOptions();

    if (changeSize && Select) {
      const options = pageSizeOptions.map((opt, i) => (
        <Select.Option
          key={i}
          value={opt.toString()}
          className={stylex(stylex.create({ fontVariantNumeric: 'tabular-nums' }))}
        >
          {this.buildOptionText(opt)}
        </Select.Option>
      ));

      changeSelect = (
        <Select
          disabled={disabled}
          prefixCls={selectPrefixCls}
          showSearch={false}
          className={`${prefixCls}-size-changer`}
          optionLabelProp="children"
          dropdownMatchSelectWidth={false}
          value={(pageSize || pageSizeOptions[0]).toString()}
          onChange={this.changeSize}
          getPopupContainer={(triggerNode: any) => triggerNode.parentNode}
        >
          {options}
        </Select>
      );
    }

    if (quickGo) {
      if (goButton) {
        gotoButton =
          typeof goButton === 'boolean' ? (
            <button
              type="button"
              onClick={this.go}
              onKeyUp={this.go}
              disabled={disabled}
              className={`${prefixCls}-quick-jumper-button`}
            >
              {locale.jump_to_confirm}
            </button>
          ) : (
            <span onClick={this.go} onKeyUp={this.go}>
              {goButton}
            </span>
          );
      }
      goInput = (
        <div className={`${prefixCls}-quick-jumper`}>
          {locale.jump_to}
          <input
            disabled={disabled}
            type="text"
            value={goInputText}
            onChange={this.handleChange}
            onKeyUp={this.go}
            onBlur={this.handleBlur}
          />
          {locale.page}
          {gotoButton}
        </div>
      );
    }

    return (
      <li className={prefixCls}>
        {changeSelect}
        {goInput}
      </li>
    );
  }
}

export default Options;
