// @universe-design/react/src/pagination/internal/Pager.tsx
import React from 'react';
import clsx from 'clsx';

interface PagerProps {
  rootPrefixCls: string;
  className?: string;
  active: boolean;
  showTitle: boolean;
  page: number;
  onClick: (page: number) => unknown;
  onKeyPress?: (ev: React.KeyboardEvent<HTMLLIElement>, onClick: (page: number) => unknown, page: number) => unknown;
  itemRender: (page: number, type: 'page', element: React.ReactElement) => React.ReactNode;
}

const Pager = (props: PagerProps) => {
  const { className } = props;
  const prefixCls = `${props.rootPrefixCls}-item`;
  const cls = clsx(
    prefixCls,
    `${prefixCls}-${props.page}`,
    {
      [`${prefixCls}-active`]: props.active,
      [`${prefixCls}-disabled`]: !props.page,
    },
    className,
  );

  const handleClick = () => {
    props.onClick(props.page);
  };

  const handleKeyPress: React.KeyboardEventHandler<HTMLLIElement> = e => {
    props.onKeyPress?.(e, props.onClick, props.page);
  };

  return (
    <li
      title={props.showTitle ? String(props.page) : undefined}
      className={cls}
      onClick={handleClick}
      onKeyPress={handleKeyPress}
      tabIndex={0}
    >
      {props.itemRender(props.page, 'page', <a rel="nofollow">{props.page}</a>)}
    </li>
  );
};

export default Pager;
