// @universe-design/react/src/table/index.tsx
import React, { useMemo, useRef, useCallback } from 'react';
import clsx from 'clsx';
import RcTable, { Summary } from '@universe-design/rc-table';
import type { TableProps as RcTableProps } from '@universe-design/rc-table/lib/Table';
import { INTERNAL_HOOKS } from '@universe-design/rc-table/lib/Table';
import stylex from '@universe-design/stylex';
import flow from 'lodash/flow';
import { cssTypographyHeadline } from '../stylex';
import type { SpinProps } from '../loading';
import { Loading } from '../loading';
import { Pagination } from '../pagination';
import { Empty } from '../empty';
import { useComponentSize, useCssScope, useLocale, useDirection } from '../config-provider/hooks';
import type { DefaultSizeType, Breakpoint, ScreenMap } from '../util';
import { isFunction, responsiveArray, getDefaultSizeAbbr, omit } from '../util';
import useSelection from './hooks/useSelection';
import type { SortState } from './hooks/useSorter';
import useSorter, { getSortData } from './hooks/useSorter';
import useFilter, { getFilterData } from './hooks/useFilter';
import useTitleColumns from './hooks/useTitleColumns';
import usePagination from './hooks/usePagination';
import { useBreakpoint } from '../grid/hooks/useBreakpoint';
import renderExpandIcon from './ExpandIcon';
import { ExpandButton } from '../shared/components/expand-button';
import type {
  Key,
  GetRowKey,
  TableLocale,
  ColumnType,
  ColumnsType,
  TableRowSelection,
  SorterResult,
  SortOrder,
  TableCurrentDataSource,
  ExpandType,
  TablePaginationConfig,
  ExpandableConfig,
  TableHeaderType,
  FilterState,
  ColumnTitleProps,
} from './interface';
import { TableAction } from './interface';
import { SELECTION_ALL, SELECTION_INVERT, DEFAULT_PAGE_SIZE } from './constants';
import { renderColumnTitle } from './utils';
import type { GroupConfig } from '@universe-design/rc-table/lib/interface';

import './style';

interface MapCache<RecordType> {
  rawData?: readonly RecordType[];
  childrenColumnName?: string;
  getRowKey?: GetRowKey<RecordType>;
  kvMap?: Map<Key, RecordType>;
}

interface ChangeEventInfo<RecordType> {
  sorter: SorterResult<RecordType> | SorterResult<RecordType>[];
  sorterStates: SortState<RecordType>[];
  filters: Record<string, (Key | boolean)[] | null>;
  filterStates: FilterState<RecordType>[];
  pagination: {
    current?: number;
    pageSize?: number;
    total?: number;
  };
  resetPagination: (() => void) | null;
}

/**
 * @docgen
 * @public
 */
export interface TableProps<RecordType>
  extends Omit<
    RcTableProps<RecordType>,
    'transformColumns' | 'internalRefs' | 'data' | 'columns' | 'scroll' | 'emptyText' | 'groupConfig' | 'rowKey'
  > {
  prefixCls?: string;
  dropdownPrefixCls?: string;
  className?: string;
  style?: React.CSSProperties;
  /**
   * 是否全边框
   */
  bordered?: boolean;
  /**
   * 表格尺寸：small | medium | large
   */
  size?: DefaultSizeType;
  /**
   * 配置默认的表头样式，默认为无底色，注意:多级表头始终为灰色底色
   * @defaultValue 'default'
   */
  headerType?: TableHeaderType;
  /**
   * 加载中
   */
  loading?: boolean | SpinProps;
  /**
   * 多语言
   */
  locale?: TableLocale;
  /**
   * 列配置
   */
  columns: ColumnsType<RecordType>;
  /**
   * 表格数据
   */
  dataSource: RcTableProps<RecordType>['data'];
  /**
   * 行是否可选择，及其配置
   */
  rowSelection?: TableRowSelection<RecordType>;
  /**
   * 是否展示分页，及其分页配置
   */
  pagination?: false | TablePaginationConfig;
  /**
   * 表格中所有浮层的渲染节点
   */
  getPopupContainer?: (triggerNode: HTMLElement) => HTMLElement;
  /**
   * 排序，也可以在 columns 中指定
   */
  sortDirections?: SortOrder[];
  /**
   * 是否显示排序提示文案，默认为 true
   */
  showSorterTooltip?: boolean;
  /**
   * 无数据时展示的缺省提示文案
   */
  emptyText?: React.ReactNode | (() => React.ReactNode);
  /**
   * 无数据时展示的内容
   */
  renderEmpty?: React.ReactNode | (() => React.ReactNode);
  /**
   * 排序、筛选、分页变化后的回调
   */
  onChange?: (
    pagination: TablePaginationConfig,
    filters: Record<string, (Key | boolean)[] | null>,
    sorter: SorterResult<RecordType> | SorterResult<RecordType>[],
    extra: TableCurrentDataSource<RecordType>,
  ) => void;
  /**
   * 表格是否可滚动，及滚动区域的宽高配置
   */
  scroll?: RcTableProps<RecordType>['scroll'] & {
    scrollToFirstRowOnChange?: boolean;
  };
  /**
   * 是否拉伸表格
   */
  stretch?: boolean;
  /**
   * 自定义行悬停状态，不会禁用触发事件
   */
  rowHoverable?: (record: RecordType, index: number, indent: number) => boolean;
  /**
   * 表格数据分组配置
   */
  groupConfig?: RcTableProps<RecordType>['groupConfig'] & {
    groupTitleRenderer?: (groupKey: Key) => React.ReactNode;
  };
  /**
   * 获取行数据对应的唯一键
   */
  rowKey?: string | GetRowKey<RecordType>;
}

/**
 * 这一层奇奇怪怪的 wrapper 是为了让 rc-table column 能取到一个字符串的 title
 * 然后加到 th[title] 里面，
 * 如果有疑问的话，需要配合 rc-table Cell 组件一起看 hhh
 */
const FakeColumnChildren = ({ children, realChildren }: React.PropsWithChildren<{ realChildren: React.ReactNode }>) => (
  <>{realChildren}</>
);

export const Table = function Table<RecordType extends Record<string, any>>(props: TableProps<RecordType>) {
  const {
    prefixCls,
    dropdownPrefixCls,
    className,
    style,
    bordered,
    headerType = 'default',
    size,
    loading,
    rowKey = 'key',
    rowClassName,
    rowSelection,
    columns,
    dataSource,
    expandable,
    pagination,
    locale = {},
    getPopupContainer,
    sortDirections,
    showSorterTooltip = true,
    emptyText,
    renderEmpty,
    stretch,
    onChange,
    rowHoverable,
    groupConfig,
  } = props;
  let { scroll } = props;

  const rcTableProps = omit(props, ['className', 'style', 'columns', 'stretch']) as TableProps<RecordType>;

  if (stretch) {
    // FIXME: x 是因为目前 fixed column 的时候 fixed layout 下 empty 会滚动，max-content 会把 layout 改为 auto，之后修好可以去掉
    // y 是因为 rc-table 里面对于传了 scroll y 的情况下会判断为 fixed header，传入 css 默认值 none，没有额外含义
    scroll = { x: 'max-content', y: 'none', ...scroll };
    rcTableProps.scroll = scroll;
  }

  const mapCacheRef = React.useRef<MapCache<RecordType>>({});

  const direction = useDirection();
  const tableCssScope = useCssScope('table');
  const prefix = prefixCls || tableCssScope();
  const dropdownCssScope = useCssScope('menu-dropdown');
  const dropdownPrefix = dropdownPrefixCls || dropdownCssScope();

  const rawData: readonly RecordType[] = useMemo(() => dataSource || [], [dataSource]);

  // columns 响应式
  const screens: ScreenMap = useBreakpoint();
  const mergedColumns: ColumnsType<RecordType> = useMemo(() => {
    const matched = new Set(responsiveArray.filter((s: Breakpoint) => Boolean(screens[s])));
    return columns.filter(
      (c: ColumnType<RecordType>) => !c.responsive || c.responsive.some((r: Breakpoint) => matched.has(r)),
    );
  }, [columns, screens]);
  const hasChildrenColumns: boolean = useMemo(() => mergedColumns.some(item => 'children' in item), [mergedColumns]);

  let mergedData: readonly RecordType[] = useMemo(() => dataSource || [], [dataSource]);

  // Size
  const defaultSize = useComponentSize();
  const mergedSize = size || defaultSize;

  // Locale
  const defaultLocale = useLocale();
  const tableLocale = { ...defaultLocale.components.table, ...locale };

  // Wrapper ClassNames
  const wrapperClassNames = clsx(`${prefix}-wrapper`, stretch && `${prefix}-wrapper-stretch`, className);

  // Loading
  let spinProps: SpinProps | undefined;
  if (typeof loading === 'boolean') {
    spinProps = {
      spinning: loading,
    };
  } else if (typeof loading === 'object') {
    spinProps = {
      ...loading,
    };
  }

  // Expand
  const mergedExpandable: ExpandableConfig<RecordType> = expandable || {};
  const { childrenColumnName = 'children' } = mergedExpandable;
  const expandType: ExpandType = React.useMemo<ExpandType>(() => {
    if (rawData.some(item => (item as RecordType & { [key: string]: RecordType[] })?.[childrenColumnName])) {
      return 'nest';
    }
    if (expandable && expandable.expandedRowRender) {
      return 'row';
    }
    return null;
  }, [rawData, expandable, childrenColumnName]);
  const defaultExpandIcon = renderExpandIcon();
  mergedExpandable.expandIcon = mergedExpandable.expandIcon || defaultExpandIcon;

  // 调整展开按钮的所在列位置
  if (expandType === 'nest' && mergedExpandable.expandIconColumnIndex === undefined) {
    mergedExpandable.expandIconColumnIndex = rowSelection ? 1 : 0;
  }

  // if (
  //   expandType === 'row' &&
  //   mergedExpandable.expandIconColumnIndex === undefined
  // ) {
  //   mergedExpandable.expandIconColumnIndex = rowSelection ? 1 : 0;
  // }

  // Indent size
  if (typeof mergedExpandable.indentSize !== 'number') {
    mergedExpandable.indentSize = 14;
  }

  // Row & Key
  const getRowKey = useMemo<GetRowKey<RecordType>>(() => {
    if (isFunction(rowKey)) {
      return rowKey;
    }
    return record => (record as RecordType & { [rowKey: string]: string })?.[rowKey] ?? '';
  }, [rowKey]);
  // get row children
  const getRowChildren = useCallback(
    (data: RecordType) => (data as RecordType & { children?: RecordType[] })?.children,
    [],
  );
  const getRecordByKey = useMemo(
    () =>
      (key: Key): RecordType => {
        if (
          !mapCacheRef.current ||
          mapCacheRef.current.rawData !== rawData ||
          mapCacheRef.current.childrenColumnName !== childrenColumnName ||
          mapCacheRef.current.getRowKey !== getRowKey
        ) {
          const kvMap = new Map<Key, RecordType>();
          const dig = (records: readonly RecordType[]) => {
            records.forEach((record, index) => {
              const rKey = getRowKey(record);
              kvMap.set(rKey, record);
              if (record && typeof record === 'object' && childrenColumnName in record) {
                dig((record as RecordType & { [key: string]: RecordType[] })[childrenColumnName] || []);
              }
            });
          };
          dig(rawData);
          mapCacheRef.current = {
            rawData,
            childrenColumnName,
            kvMap,
            getRowKey,
          };
        }

        return mapCacheRef.current.kvMap!.get(key)!;
      },
    [rawData, childrenColumnName, getRowKey],
  );

  // Group
  const groupTitleRenderer = groupConfig?.groupTitleRenderer;
  const mergedGroupConfig: GroupConfig<RecordType> = groupConfig || {};

  const groupRowRenderer = useMemo(() => {
    if (typeof mergedGroupConfig.groupRowRenderer !== 'function') {
      return (groupKey: Key, expanded: boolean) => {
        const title = typeof groupTitleRenderer === 'function' ? groupTitleRenderer(groupKey) : groupKey;
        return (
          <div className={`${prefix}-group-title`}>
            <ExpandButton
              className={`${prefix}-group-title-expand-button`}
              isExpanded={expanded}
              iconType="ExpandDownFilled"
              readonly={true}
            />
            {title}
          </div>
        );
      };
    }

    return mergedGroupConfig.groupRowRenderer;
  }, [groupTitleRenderer, mergedGroupConfig.groupRowRenderer, prefix]);

  mergedGroupConfig.groupRowRenderer = groupRowRenderer;

  // Events
  const internalRefs = {
    body: useRef<HTMLDivElement>(),
  };
  const changeEventInfo: Partial<ChangeEventInfo<RecordType>> = {};
  const triggerOnChange = (info: Partial<ChangeEventInfo<RecordType>>, action: TableAction, reset = false) => {
    const changeInfo = {
      ...changeEventInfo,
      ...info,
    };

    if (reset) {
      if (isFunction(changeEventInfo.resetPagination)) {
        changeEventInfo.resetPagination!();
      }

      // Reset event param
      if (changeInfo.pagination!.current) {
        changeInfo.pagination!.current = 1;
      }

      // Trigger pagination events
      if (pagination && pagination.onChange) {
        pagination.onChange(1, changeInfo.pagination!.pageSize);
      }
    }

    if (scroll && scroll.scrollToFirstRowOnChange !== false && internalRefs.body.current) {
      internalRefs.body.current.scrollTo({
        top: 0,
        behavior: 'smooth',
      });
    }

    if (onChange) {
      onChange(changeInfo.pagination!, changeInfo.filters!, changeInfo.sorter!, {
        currentDataSource: getFilterData(
          getSortData(rawData, changeInfo.sorterStates!, childrenColumnName),
          changeInfo.filterStates!,
        ),
        action,
      });
    }
  };

  // wrap column
  const wrapTitle = useCallback(
    (innerColumns: ColumnsType<RecordType>, wrapName: string, otherClassName: string): ColumnsType<RecordType> =>
      innerColumns.map(column => {
        let newColumn: ColumnsType<RecordType>[number] = column;
        newColumn = {
          ...newColumn,
          title: (renderProps: ColumnTitleProps<RecordType>) => (
            <span className={clsx(`${prefix}-column-${wrapName}`, otherClassName)}>
              {renderColumnTitle(column.title, renderProps)}
            </span>
          ),
        };
        if ('children' in newColumn) {
          newColumn = {
            ...newColumn,
            children: wrapTitle(newColumn.children, wrapName, otherClassName),
          };
        }
        return newColumn;
      }),
    [prefix],
  );

  // Sorter
  const onSorterChange = (
    sorter: SorterResult<RecordType> | SorterResult<RecordType>[],
    sorterStates: SortState<RecordType>[],
  ) => {
    triggerOnChange({ sorter, sorterStates }, TableAction.SORT, false);
  };
  const [transformSorterColumns, sortStates, sorterTitleProps, getSorters] = useSorter<RecordType>({
    prefix,
    mergedColumns,
    onSorterChange,
    sortDirections: sortDirections || ['ascend', 'descend'],
    tableLocale,
    showSorterTooltip,
  });
  mergedData = React.useMemo(
    () => getSortData(rawData, sortStates, childrenColumnName),
    [rawData, sortStates, childrenColumnName],
  );
  changeEventInfo.sorter = getSorters();
  changeEventInfo.sorterStates = sortStates;

  // Filter
  const onFilterChange = (
    filters: Record<string, (Key | boolean)[] | null>,
    filterStates: FilterState<RecordType>[],
  ) => {
    triggerOnChange({ filters, filterStates }, TableAction.FILTER, true);
  };
  const [transformFilterColumns, filterStates, getFilters] = useFilter<RecordType>({
    prefix,
    dropdownPrefix,
    tableLocale,
    mergedColumns,
    onFilterChange,
    getPopupContainer,
  });
  mergedData = getFilterData(mergedData, filterStates);
  changeEventInfo.filters = getFilters();
  changeEventInfo.filterStates = filterStates;

  // Pagination
  const onPaginationChange = (param: { current?: number; pageSize?: number; total?: number }) => {
    triggerOnChange({ pagination: param }, TableAction.PAGINATION, false);
  };
  const [mergedPagination, resetPagination, getPaginationParam] = usePagination(
    mergedData.length,
    pagination,
    onPaginationChange,
  );
  changeEventInfo.pagination = pagination === false ? {} : getPaginationParam();
  changeEventInfo.resetPagination = resetPagination;

  // Render Pagination
  let topPaginationNode: React.ReactNode = null;
  let bottomPaginationNode: React.ReactNode = null;
  if (pagination !== false && mergedPagination.total) {
    // 分页器大小
    const paginationSize: TablePaginationConfig['size'] = mergedPagination.size
      ? mergedPagination.size
      : mergedSize === 'large'
      ? 'medium'
      : mergedSize;

    // 分页器 dom
    const renderPagination = (position: string) => (
      <Pagination
        className={`${prefix}-pagination ${prefix}-pagination-${position}`}
        {...mergedPagination}
        size={paginationSize}
      />
    );

    // 分页器位置
    const defaultPosition = direction === 'rtl' ? 'left' : 'right';
    if (mergedPagination.position !== null && Array.isArray(mergedPagination.position)) {
      const topPos = mergedPagination.position.find(p => p.indexOf('top') !== -1);
      const bottomPos = mergedPagination.position.find(p => p.indexOf('bottom') !== -1);
      if (!topPos && !bottomPos) {
        bottomPaginationNode = renderPagination(defaultPosition);
      } else {
        if (topPos) {
          topPaginationNode = renderPagination(topPos.toLowerCase().replace('top', ''));
        }
        if (bottomPos) {
          bottomPaginationNode = renderPagination(bottomPos.toLowerCase().replace('bottom', ''));
        }
      }
    } else {
      bottomPaginationNode = renderPagination(defaultPosition);
    }
  }

  // 分页数据
  const pageData = useMemo<readonly RecordType[]>(() => {
    if (pagination === false || !mergedPagination.pageSize) {
      return mergedData;
    }

    const { current = 1, total = 0, pageSize = DEFAULT_PAGE_SIZE } = mergedPagination;

    if (mergedData.length < total) {
      if (mergedData.length > pageSize) {
        return mergedData.slice((current - 1) * pageSize, current * pageSize);
      }
      return mergedData;
    }

    return mergedData.slice((current - 1) * pageSize, current * pageSize);
  }, [mergedData, mergedPagination, pagination]);

  // Selections
  const [transformSelectionColumns, selectedKeySet] = useSelection<RecordType>(rowSelection, {
    prefix,
    data: mergedData,
    pageData,
    getRowKey,
    getRowChildren,
    getRecordByKey,
    expandType,
    childrenColumnName,
    tableLocale,
    getPopupContainer,
  });

  const internalRowClassName = (record: RecordType, index: number, indent: number) => {
    let mergedRowClassName;
    if (isFunction(rowClassName)) {
      mergedRowClassName = clsx(rowClassName(record, index, indent));
    } else {
      mergedRowClassName = clsx(rowClassName);
    }

    if (!rowHoverable || rowHoverable(record, index, indent)) {
      mergedRowClassName = clsx(`${prefix}-row-hoverable`, mergedRowClassName);
    }

    return clsx(
      {
        [`${prefix}-row-selected`]: selectedKeySet.has(getRowKey(record)),
      },
      mergedRowClassName,
    );
  };

  // Title
  const columnTitleProps = React.useMemo(() => ({ ...sorterTitleProps }), [sorterTitleProps]);
  const [transformTitleColumns] = useTitleColumns<RecordType>(columnTitleProps);
  const renderEmptyCmp = React.useMemo(
    () => (
      <div className={clsx(`${prefix}-empty-wrapper`)}>
        {renderEmpty || <Empty description={emptyText || locale?.emptyText} />}
      </div>
    ),
    [emptyText, locale?.emptyText, prefix, renderEmpty],
  );

  const usedSize = getDefaultSizeAbbr(mergedSize);

  const finalColumns = React.useMemo(() => {
    const wrapTitleColumns = (_columns: ColumnsType<RecordType>) =>
      wrapTitle(_columns, 'title', stylex(cssTypographyHeadline));

    const wrapContainerColumns = (_columns: ColumnsType<RecordType>) => wrapTitle(_columns, 'wrapper', '');

    const wrapFakeChildrenColumns = (_columns: ColumnsType<RecordType>) =>
      _columns.map((column, index) => ({
        ...column,
        title: (
          <FakeColumnChildren realChildren={column.title}>
            {mergedColumns[rowSelection ? index - 1 : index]?.title}
          </FakeColumnChildren>
        ),
      }));

    /**
     * 会依次执行以下函数，封装成最终的 column title
     */
    const composeFn = [
      // 给 title 加上 一个 span 包裹
      wrapTitleColumns,
      // 添加 排序的 dom 结构
      transformSorterColumns,
      // 封装上筛选功能
      transformFilterColumns,
      // 整体再套一个 div 壳子
      wrapContainerColumns,
      // 加上 selection 的 checkbox
      transformSelectionColumns,
      // 执行一下函数
      transformTitleColumns,
      // 封装上假的 title
      wrapFakeChildrenColumns,
    ];
    return flow(composeFn)(mergedColumns) as ColumnsType<RecordType>;
  }, [
    mergedColumns,
    rowSelection,
    transformFilterColumns,
    transformSelectionColumns,
    transformSorterColumns,
    transformTitleColumns,
    wrapTitle,
  ]);

  return (
    <div className={wrapperClassNames} style={style}>
      <Loading spinning={false} {...spinProps}>
        {loadingCtx => (
          <>
            {topPaginationNode}
            <RcTable<RecordType>
              {...rcTableProps}
              columns={finalColumns}
              expandable={mergedExpandable}
              prefixCls={prefix}
              className={clsx({
                [`${prefix}-${usedSize}`]: usedSize,
                [`${prefix}-bordered`]: bordered,
                [`${prefix}-empty`]: rawData.length === 0,
                [`${prefix}-multi-header`]: hasChildrenColumns,
                [`${prefix}-${headerType}-header`]: headerType !== 'default',
                [`${prefix}-stretch`]: stretch,
              })}
              data={pageData}
              rowKey={getRowKey}
              rowClassName={internalRowClassName}
              emptyText={loadingCtx.loading ? <div className={`${prefix}-empty-loading`} /> : renderEmptyCmp}
              groupConfig={mergedGroupConfig}
              // Internal
              // 后期如果 rc-table 升级了，不建议使用 internal
              // 这里必须使用者 internal 参数，才能拿到内部的 scrollBodyRef，实现 scrollToFirstRowOnChange
              internalHooks={INTERNAL_HOOKS}
              internalRefs={
                internalRefs as {
                  body: React.MutableRefObject<HTMLDivElement>;
                }
              }
            />
            {bottomPaginationNode}
          </>
        )}
      </Loading>
    </div>
  );
};

Table.SELECTION_ALL = SELECTION_ALL;
Table.SELECTION_INVERT = SELECTION_INVERT;
Table.Summary = Summary;

export * from './interface';
