// @universe-design/react/src/table/interface.ts
import type React from 'react';
import type {
  DefaultRecordType,
  ColumnType as RcColumnType,
  RenderedCell as RcRenderedCell,
  ExpandableConfig,
  DataIndex,
} from '@universe-design/rc-table/lib/interface';
import type { CheckboxProps } from '../checkbox/Checkbox';
import type { PaginationProps } from '../pagination/Pagination';
import type { Breakpoint } from '../util';

export type { DefaultRecordType, ExpandableConfig };

export type Key = React.Key;

export type GetRowKey<RecordType> = (record: RecordType) => Key;

export type CompareFn<T> = (a: T, b: T, sortOrder?: SortOrder) => number;

// header
export type TableHeaderType = 'default' | 'highlight';

// Locale
export interface TableLocale {
  filterTitle?: string;
  filterConfirm?: React.ReactNode;
  filterReset?: React.ReactNode;
  filterEmptyText?: React.ReactNode;
  emptyText?: React.ReactNode | (() => React.ReactNode);
  selectAll?: React.ReactNode;
  selectInvert?: React.ReactNode;
  selectionAll?: React.ReactNode;
  sortTitle?: string;
  expand?: string;
  collapse?: string;
  triggerDesc?: string;
  triggerAsc?: string;
  cancelSort?: string;
}

// Sort
export type SortOrder = 'descend' | 'ascend' | null;
export interface SorterResult<RecordType> {
  column?: ColumnType<RecordType>;
  order?: SortOrder;
  field?: DataIndex;
  columnKey?: Key;
}
export enum TableSort {
  ASCEND = 'ascend',
  DESCEND = 'descend',
}

// Filter
export type FilterKey = Key | boolean;
export interface ColumnFilterItem {
  text: React.ReactNode;
  value: FilterKey;
  children?: ColumnFilterItem[];
}
export interface FilterDropdownProps {
  prefixCls: string;
  setSelectedKeys: (selectedKeys: FilterKey[]) => void;
  selectedKeys: FilterKey[];
  confirm: () => void;
  clearFilters?: () => void;
  filters?: ColumnFilterItem[];
  visible: boolean;
}
export interface FilterState<RecordType> {
  column: ColumnType<RecordType>;
  key: Key;
  filteredKeys?: FilterKey[] | null;
  forceFiltered?: boolean;
}

export type FilterIconAlignType = 'left' | 'right';

// Expand
export type ExpandType = null | 'row' | 'nest';

// Selection
export type TableSelectionItem = SelectionItem;
export type SelectionSelectFn<T> = (record: T, selected: boolean, selectedRows: T[], nativeEvent: Event) => void;
export type SelectionItemSelectFn = (currentRowKeys: Key[]) => void;
export interface SelectionItem {
  key: string;
  text: React.ReactNode;
  onSelect?: SelectionItemSelectFn;
}
export type RowSelectionType = 'checkbox' | 'radio';
export interface TableRowSelection<T> {
  /** Keep the selection keys in list even the key not exist in `dataSource` anymore */
  preserveSelectedRowKeys?: boolean;
  type?: RowSelectionType;
  selectedRowKeys?: Key[];
  onChange?: (selectedRowKeys: Key[], selectedRows: T[]) => void;
  getCheckboxProps?: (record: T) => Partial<Omit<CheckboxProps, 'checked' | 'defaultChecked'>>;
  onSelect?: SelectionSelectFn<T>;
  /** @deprecated This function is meaningless and should use `onChange` instead */
  onSelectAll?: (selected: boolean, selectedRows: T[], changeRows: T[]) => void;
  /** @deprecated This function is meaningless and should use `onChange` instead */
  onSelectInvert?: (selectedRowKeys: Key[]) => void;
  selections?: TableSelectionItem[] | boolean;
  hideSelectAll?: boolean;
  fixed?: boolean;
  columnWidth?: string | number;
  columnTitle?: string | React.ReactNode;
  checkStrictly?: boolean;
  renderCell?: (
    value: boolean,
    record: T,
    index: number,
    originNode: React.ReactNode,
  ) => React.ReactNode | RcRenderedCell<T>;
}

// Column & Columns
export interface ColumnType<RecordType> extends RcColumnType<RecordType> {
  title?: ColumnTitle<RecordType>;

  sorter?:
    | boolean
    | CompareFn<RecordType>
    | {
        compare?: CompareFn<RecordType>;
        multiple?: number;
      };
  sortOrder?: SortOrder;
  defaultSortOrder?: SortOrder;
  sortDirections?: SortOrder[];
  showSorterTooltip?: boolean;

  filtered?: boolean;
  filters?: ColumnFilterItem[];
  filterDropdown?: React.ReactNode | ((props: FilterDropdownProps) => React.ReactNode);
  filterMultiple?: boolean;
  filteredValue?: Key[] | null;
  defaultFilteredValue?: Key[] | null;
  filterIcon?: React.ReactNode | ((filtered: boolean) => React.ReactNode);
  onFilter?: (value: FilterKey, record: RecordType) => boolean;
  filterDropdownVisible?: boolean;
  onFilterDropdownVisibleChange?: (visible: boolean) => void;
  filterIconAlign?: FilterIconAlignType;

  responsive?: Breakpoint[];
}
export interface ColumnTitleProps<RecordType> {
  sortColumns?: { column: ColumnType<RecordType>; order: SortOrder }[];
  filters?: Record<string, string[]>;
}
export type ColumnTitle<RecordType> = React.ReactNode | ((props: ColumnTitleProps<RecordType>) => React.ReactNode);
export interface ColumnGroupType<RecordType> extends Omit<ColumnType<RecordType>, 'dataIndex'> {
  children: ColumnsType<RecordType>;
}
export type ColumnsType<RecordType> = (ColumnGroupType<RecordType> | ColumnType<RecordType>)[];
export type TransformColumns<RecordType> = (columns: ColumnsType<RecordType>) => ColumnsType<RecordType>;

// Other
export type GetPopupContainerInTable = (triggerNode: HTMLElement) => HTMLElement;

export enum TableAction {
  PAGINATION = 'pagination',
  SORT = 'sort',
  FILTER = 'filter',
}

export interface TableCurrentDataSource<RecordType> {
  currentDataSource: readonly RecordType[];
  action: TableAction;
}

export type GetCheckDisabled<RecordType> = (record: RecordType) => boolean;

export type TablePaginationPosition =
  | 'topLeft'
  | 'topCenter'
  | 'topRight'
  | 'bottomLeft'
  | 'bottomCenter'
  | 'bottomRight';
export interface TablePaginationConfig extends PaginationProps {
  position?: TablePaginationPosition[];
}
