// @universe-design/react/src/table/utils.ts
import type React from 'react';
import type { ColumnType, ColumnTitle, ColumnTitleProps, Key, GetRowKey, FilterKey } from './interface';

import type { TreeNodeProps } from '../tree/interface';
import { CheckedType } from '../tree/interface';
import type { TKeyValueMap } from '../util';
import { isArray } from '../util';

export { CheckedType };

export function getColumnKey<RecordType>(column: ColumnType<RecordType>, defaultKey: string): Key {
  if ('key' in column && column.key !== undefined && column.key !== null) {
    return column.key;
  }
  if (column.dataIndex) {
    return isArray(column.dataIndex) ? column.dataIndex.join('.') : column.dataIndex;
  }
  return defaultKey;
}

export function getColumnPos(index: number, pos?: string) {
  return pos ? `${pos}-${index}` : `${index}`;
}

export function renderColumnTitle<RecordType>(title: ColumnTitle<RecordType>, props: ColumnTitleProps<RecordType>) {
  if (typeof title === 'function') {
    return title(props);
  }
  return title;
}

export interface Entity {
  node: React.ReactElement<TreeNodeProps>;
  index: number;
  key: Key;
  parent?: Entity;
  children?: Entity[];
}

interface DataNode {
  children?: DataNode[];
}

interface DataEntity extends Omit<Entity, 'node' | 'parent' | 'children'> {
  node: DataNode;
  parent?: DataEntity;
  children?: DataEntity[];
  depth: number;
  value: Key;
}

export function traverseDataNodes(
  data: readonly DataNode[],
  callback: (param: { node: DataNode; index: number; key: Key; parentKey: Key; depth: number }) => void,
  getRowKey: GetRowKey<DataNode>,
) {
  function processNode(
    node: DataNode | null,
    index: number,
    parent?: { node: DataNode | null; key: Key | undefined; depth: number },
  ) {
    const children = node ? node.children : data;

    // Process node if is not root
    if (node) {
      const key: Key = getRowKey(node);
      const obj = {
        node,
        index,
        key,
        parentKey: parent!.node ? getRowKey(parent!.node) : '',
        depth: parent!.depth + 1,
      };

      callback(obj);
    }

    // Process children node
    if (children) {
      children.forEach((subNode, subIndex) => {
        processNode(subNode, subIndex, {
          node,
          key: getRowKey(node!),
          depth: parent ? parent.depth + 1 : -1,
        });
      });
    }
  }

  processNode(null, 0);
}

// TODO 这里类型不太对，要再仔细查一下
export function convertDataToEntities<RecordType>(data: readonly RecordType[], getRowKey: GetRowKey<RecordType>) {
  const keyEntities: { [key: string]: DataEntity } = {};
  const wrapper = {
    keyEntities,
  };

  traverseDataNodes(
    data as readonly DataEntity[],
    item => {
      const { node, index, key, parentKey, depth } = item;
      const entity: DataEntity = { node, index, key, depth, value: key };

      // const mergedKey = getKey(key, pos);

      keyEntities[key] = entity;

      // Fill children
      entity.parent = keyEntities[parentKey];
      if (entity.parent) {
        entity.parent.children = entity.parent.children || [];
        entity.parent.children.push(entity);
      }
    },
    getRowKey as GetRowKey<DataNode>,
  );

  return wrapper;
}

export type CheckedTypeCache = Map<Key, CheckedType>;

export function isNonLeaf2(node: DataEntity): node is DataEntity & Required<Pick<DataEntity, 'children'>> {
  return Boolean(node.children && node.children.length !== 0);
}

export function getCheckedType(node: DataEntity, checkedSet: TKeyValueMap<boolean>, cache: CheckedTypeCache) {
  const key = node.value;
  let checkedType = cache.get(key);
  if (checkedType !== undefined) {
    return checkedType;
  }
  if (
    checkedSet[key] ||
    (isNonLeaf2(node) && node.children.every(_node => getCheckedType(_node, checkedSet, cache) === CheckedType.Checked))
  ) {
    checkedType = CheckedType.Checked;
    // 将所有的子元素都设置为 checked
    for (let stack: DataEntity[] = [node]; ; ) {
      const current = stack.pop();
      if (!current) {
        break;
      }
      cache.set(current.key, CheckedType.Checked);
      stack.push(...(current.children || []));
    }
  } else if (
    !isNonLeaf2(node) ||
    node.children.every(_node => getCheckedType(_node, checkedSet, cache) === CheckedType.Unchecked)
  ) {
    checkedType = CheckedType.Unchecked;
    cache.set(key, CheckedType.Unchecked);
  } else {
    checkedType = CheckedType.Indeterminate;
    cache.set(key, CheckedType.Indeterminate);
  }
  return checkedType;
}

export function checkedSet2Cache(keyEntities: { [key: string]: DataEntity }, checkedSet: TKeyValueMap<boolean>) {
  const cache: CheckedTypeCache = new Map();
  for (const node of Object.values(keyEntities)) {
    getCheckedType(node, checkedSet, cache);
  }
  return cache;
}

function updateUpLevel<RecordType>(_node: DataEntity, cache: CheckedTypeCache, getRowKey: GetRowKey<RecordType>) {
  const { parent } = _node;
  const siblings = parent?.children;
  if (parent && siblings) {
    const parentKey = parent.value;
    const isSiblingsAllChecked = siblings.every(o => {
      const key = getRowKey(o as unknown as RecordType) as string;
      return cache.get(key) === CheckedType.Checked;
    });
    const isSiblingsAllUnchecked = siblings.every(o => {
      const key = o.value;
      return cache.get(key) === CheckedType.Unchecked;
    });
    if (isSiblingsAllChecked) {
      cache.set(parentKey, CheckedType.Checked);
    } else if (isSiblingsAllUnchecked) {
      cache.set(parentKey, CheckedType.Unchecked);
    } else {
      cache.set(parentKey, CheckedType.Indeterminate);
    }
    updateUpLevel(parent, cache, getRowKey);
  }
}

function updateDownLevel<RecordType>(
  _node: DataEntity,
  cache: CheckedTypeCache,
  type: CheckedType,
  getRowKey: GetRowKey<RecordType>,
) {
  if (isNonLeaf2(_node)) {
    _node.children.forEach(o => {
      const key = o.value;
      cache.set(key, type);
      updateDownLevel(o, cache, type, getRowKey);
    });
  }
}

export function check<RecordType>(
  _node: DataEntity,
  cache: CheckedTypeCache,
  checkStrictly: boolean,
  getRowKey: GetRowKey<RecordType>,
) {
  cache.set(_node.value, CheckedType.Checked);
  if (!checkStrictly) {
    updateUpLevel(_node, cache, getRowKey);
    updateDownLevel(_node, cache, CheckedType.Checked, getRowKey);
  }
}

export function unCheck<RecordType>(
  _node: DataEntity,
  cache: CheckedTypeCache,
  checkStrictly: boolean,
  getRowKey: GetRowKey<RecordType>,
) {
  cache.set(_node.value, CheckedType.Unchecked);
  if (!checkStrictly) {
    updateUpLevel(_node, cache, getRowKey);
    updateDownLevel(_node, cache, CheckedType.Unchecked, getRowKey);
  }
}

export function cache2CheckedSet(keyEntities: { [key: string]: DataEntity }, cache: CheckedTypeCache) {
  const checkedSet: TKeyValueMap<boolean> = {};
  for (const node of Object.values(keyEntities)) {
    const checkedType = cache.get(node.value);
    if (checkedType === undefined) {
      throw new Error(
        `checkedType of ${node.value} is not found in cache, are you sure that you have called checkedSet2Cache brefore`,
      );
    }
    checkedSet[node.value] = checkedType === CheckedType.Checked;
  }
  return checkedSet;
}

export const selectWithShift = (
  lastSelectedKey: Key | null,
  curSelectedKey: Key,
  allKeys: Key[],
  preSelectedKeySet: Set<Key>,
) => {
  let startIndex = -1;
  let endIndex = -1;
  const selectedToUnselected = preSelectedKeySet.has(curSelectedKey);
  const curSelectedKeySet = new Set<Key>(preSelectedKeySet);

  const boundaryKeys = new Set([lastSelectedKey, curSelectedKey]);
  allKeys.some((recordKey, recordIndex) => {
    if (boundaryKeys.has(recordKey)) {
      if (startIndex === -1) {
        startIndex = recordIndex;
      } else {
        endIndex = recordIndex;
        return true;
      }
    }
    return false;
  });

  if (startIndex !== endIndex && endIndex !== -1) {
    // shift 批量选择多个 key:[startIndex, endIndex]
    const selectedRangedKeys = allKeys.slice(startIndex, endIndex + 1);

    if (selectedToUnselected) {
      // 取消选择范围内的 key
      selectedRangedKeys.forEach(k => {
        if (curSelectedKeySet.has(k)) {
          curSelectedKeySet.delete(k);
        }
      });
    } else {
      // 选择范围内的 key
      selectedRangedKeys.forEach(k => {
        if (!curSelectedKeySet.has(k)) {
          curSelectedKeySet.add(k);
        }
      });
    }
  } else {
    if (selectedToUnselected) {
      curSelectedKeySet.delete(curSelectedKey);
    } else {
      curSelectedKeySet.add(curSelectedKey);
    }
  }
  return curSelectedKeySet;
};

// filter
export const serializeFilterKey = (key: FilterKey): string => JSON.stringify(key);

export const serializeFilterKeys = (selectedKeys: FilterKey[]) =>
  selectedKeys.map(filterKey => serializeFilterKey(filterKey));

export const unSerializeFilterKeys = (serializedFilterKeys: string[]): FilterKey[] =>
  serializedFilterKeys.map(serializedKey => JSON.parse(serializedKey));
