// @universe-design/react/src/transfer/stories/homeless.stories.tsx
import * as React from 'react';
import type { TransferOption } from '@universe-design/react';
import {
  Transfer,
  TransferSourcePanel,
  TransferSourcePanelHeader,
  TransferSourceList,
  TransferTargetPanel,
  TransferTargetPanelHeader,
  TransferTargetList,
  TransferTargetOption,
  TransferTargetOptionLabel,
  TransferTargetOptionRemoveButton,
  Slider,
  InputNumber,
} from '@universe-design/react';

export default {
  title: 'Components/Transfer/临时示例',
  component: Transfer,
};

const options: TransferOption[] = [...new Array(100)].map((item, index) => ({
  value: `${index}`,
  label: new Array((index % 8) + 1).fill('option').join('-'),
  disabled: index % 3 === 0,
}));

export const OverflowAndTooltip = () => (
  <Transfer options={options} defaultValue={['0', '1', '2'] as string[]} style={{ width: 600 }} />
);

export const NumberValue = () => (
  <Transfer
    style={{ maxWidth: 600 }}
    onChange={(v, option) => {
      console.log(v, option);
    }}
    options={[
      { value: 1, label: '1' },
      { value: 4, label: '1' },
    ]}
  />
);

export const WithRepeatKeys = () => (
  <>
    <Transfer
      style={{ maxWidth: 600 }}
      onChange={(v, option) => {
        console.log(v, option);
      }}
      options={[
        { value: '1', label: '1-1' },
        { value: '1', label: '1-2' },
      ]}
      defaultValue={['1'] as string[]}
    />
    <Transfer
      style={{ maxWidth: 600 }}
      onChange={(v, option) => {
        console.log(v, option);
      }}
      defaultValue={['1']}
      options={[
        {
          title: '分组1',
          key: 'group',
          children: [{ value: '1', label: '1-1' }],
        },
        {
          title: '分组2',
          key: 'group',
          children: [{ value: '1', label: '2-1' }],
        },
        {
          title: '分组3',
          key: 'group',
          children: [{ value: '1', label: '3-1' }],
        },
      ]}
      group
    />
  </>
);

export const DifferentLayoutAndDividerTest = () => {
  const [count, setCount] = React.useState(10);
  const [value, setValue] = React.useState(300);
  const _options = React.useMemo(
    () =>
      Array(count)
        .fill(0)
        .map((_, i) => ({
          value: i,
        })),
    [count],
  );
  return (
    <>
      <InputNumber
        value={count}
        min={0}
        max={20}
        onChange={val => {
          if (val === undefined) {
            return;
          }
          setCount(val);
        }}
      />
      <Slider min={150} max={450} step={10} value={value} onChange={val => setValue(val)} style={{ width: 600 }} />
      <Transfer style={{ maxWidth: 600, height: value }} options={options} />
      <Transfer style={{ maxWidth: 600, height: value, marginTop: 16 }} options={_options}>
        <TransferSourcePanel>
          <TransferSourcePanelHeader />
          <TransferSourceList />
        </TransferSourcePanel>
        <TransferTargetPanel
          optionRenderer={() => (
            <TransferTargetOption after={<TransferTargetOptionRemoveButton />}>
              <TransferTargetOptionLabel />
            </TransferTargetOption>
          )}
        >
          <TransferTargetPanelHeader />
          <TransferTargetList />
        </TransferTargetPanel>
      </Transfer>
    </>
  );
};
