// @universe-design/react/src/transfer/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Basic from '../demo/0.basic';
import C1Group from '../demo/1.group';
import C2OptionRender from '../demo/2.optionRender';
import C4Search from '../demo/4.search';
import C5Loading from '../demo/5.loading';
import C6Tree from '../demo/6.tree';
import CCompose from '../demo/compose';
if (!(C0Basic as Story).storyName) {
  (C0Basic as Story).storyName = '0. 基础用法';
  (C0Basic as Story).parameters = {
    docs: {
      description: {
        story: '基本用法。\n\n',
      },
    },
  };
}
if (!(C1Group as Story).storyName) {
  (C1Group as Story).storyName = '1. 分组';
  (C1Group as Story).parameters = {
    docs: {
      description: {
        story:
          '当添加 group 属性时，会显示分组信息。 `group` 也支持传递一个数组，分别设置 Source 和 Target 面板是否显示分组信息。\n\n',
      },
    },
  };
}
if (!(C2OptionRender as Story).storyName) {
  (C2OptionRender as Story).storyName = '2. 自定义选项渲染';
  (C2OptionRender as Story).parameters = {
    docs: {
      description: {
        story: '自定义选项渲染, 在下面的例子中，实现选项使用两行文本显示。\n\n',
      },
    },
  };
}
if (!(C4Search as Story).storyName) {
  (C4Search as Story).storyName = '4. 远程搜索';
  (C4Search as Story).parameters = {
    docs: {
      description: {
        story:
          '这里实现一个远程搜索的示例。该示例演示了如何根据搜索条件远程拉取数据，以及如何设置 `autoFocus` 及默认搜索条件。\n',
      },
    },
  };
}
if (!(C5Loading as Story).storyName) {
  (C5Loading as Story).storyName = '5. 触底加载';
  (C5Loading as Story).parameters = {
    docs: {
      description: {
        story: '这里给出一个实现类似触底加载效果的实现示例。\n\n',
      },
    },
  };
}
if (!(C6Tree as Story).storyName) {
  (C6Tree as Story).storyName = '6. 自定义面板渲染';
  (C6Tree as Story).parameters = {
    docs: {
      description: {
        story: '支持自定义面板内容，往往需要和 useTransferContext 一起使用。\n\n',
      },
    },
  };
}
if (!(CCompose as Story).storyName) {
  (CCompose as Story).storyName = '7. 组合式用法';
  (CCompose as Story).parameters = {
    docs: {
      description: {
        story:
          '组合式用法。我们导出了 Transfer 的所有"部件"，供用户自己组装使用。你可以基于此实现隐藏部分部件，或插入、替换自己的内容。\n\n```\nTransfer\n    TransferSourcePanel\n        TransferSourcePanelHeader\n            TransferSearch\n            TransferSelectAll\n        TransferSourceList\n    TransferTargetPanel\n        TransferTargetPanelHeader\n            TranferSelectionSummary\n        TransferTargetList\n```\n\n在下面这个例子中，我们隐藏了【选择全部】(即 `TransferSelectAll`) 组件，以及添加了一块自定义区域。\n',
      },
    },
  };
}

export default {
  title: 'Components/Transfer/官网示例',
  tags: ['autodocs'],
};

export { C0Basic, C1Group, C2OptionRender, C4Search, C5Loading, C6Tree, CCompose };
