// @universe-design/react/src/tree-select/index.tsx
/* eslint-disable react-hooks/rules-of-hooks */
/* eslint-disable max-lines-per-function */
import type { ComponentPropsWithRef, Context, FC, PropsWithChildren, ReactElement, Ref, RefAttributes } from 'react';
import React, { useCallback, useContext, useMemo } from 'react';
import { COMPONENT_VERSION_ID, useCssScope } from '../config-provider';
import { SelectActionType } from '../select/internal.interface';
import { SelectContext, selectGenerator } from '../select';
import { defaultFilterTreeNode } from './common';
import { ShowCheckedStrategy, traverseTreeConfig, TreeNode, getTreeNodeConfigPath } from '../tree';
import { genArray, isFunction, isNonNullable } from '../util';
import { treeChildren2TreeNodes } from '../tree/common';
import { genDoublyTreeNodeMap } from '../tree/util';
import { TreeCheckedProvider } from '../tree/hooks/TreeCheckedContext';
import { strategyMap } from '../tree/hooks/checkStrategy';
import { useTreeCheck } from '../tree/hooks/useTreeCheck';
import { useTree } from '../tree/hooks/useTree';
import { TreeContainer } from '../tree/TreeContainer';
import { TreeContextProvider } from '../tree/hooks/TreeContext';
import { SelectOverlayContainer } from '../select/SelectOverlayContainer';
import type { CheckStrictlyKeys, TreeNodeConfig, TreeRef, Tree } from '../tree';
import type {
  MultipleTreeSelectProps,
  SingleTreeSelectProps,
  TreeSelectOverlayProps,
  TreeSelectProps,
} from './interface';
import type { RefSelectRef, SelectContextType } from '../select';
import type {
  MultipleInternalSelectProps,
  SingleInternalSelectProps,
  InternalSelectProps,
  SelectAction,
} from '../select/internal.interface';

import './style';

const RefSelect = selectGenerator<string, TreeNodeConfig>({
  selectorCssScope: 'select',
  dropdownCssScope: 'select',
  defaultOptionLabelProp: option => option.label ?? option.title ?? option.value,
});

/**
 *   因为 tree 和 select-tree的value参数不一样， 所以需要将 tree-select的value 格式化为tree的value
 *   string[] -> string[] | {checked: string[], unchecked: string[]}
 */

function toCheckStrictlyValue(value: string | string[] | undefined): CheckStrictlyKeys {
  if (Array.isArray(value)) {
    return { checked: value, halfChecked: [] } as CheckStrictlyKeys;
  }
  return { checked: [], halfChecked: [] };
}

const TreeSelectOverlay: FC<
  TreeSelectOverlayProps & {
    multiple: boolean;
    treeCheckable: boolean;
    treeRef: React.Ref<TreeRef>;
  }
> = ({
  multiple,
  treeCheckable,
  treeData,
  treeDefaultExpandAll,
  treeDefaultExpandedKeys,
  treeExpandedKeys,
  onTreeExpand,
  listHeight,
  filterTreeNode: filterTreeNodeProp,
  showCheckedStrategy,
  treeCheckStrictly,
  switcherIcon,
  loadData,
  virtual,
  nodeRenderer,
  treeRef,
  treeLoadedKeys,
  onTreeLoad,
}) => {
  const cssScope = useCssScope('treeSelect');
  const { inputValue, valueSet, set } = useContext(SelectContext as Context<SelectContextType<string, TreeNodeConfig>>);

  let treeProps: ComponentPropsWithRef<typeof Tree> = {
    ref: treeRef,
    searchValue: inputValue,
    defaultExpandAll: treeDefaultExpandAll,
    expandedKeys: treeExpandedKeys,
    defaultExpandedKeys: treeDefaultExpandedKeys,
    onExpand: onTreeExpand,
    multiple,
    selectable: !treeCheckable,
    checkable: treeCheckable,
    showChecker: !treeCheckable,
    treeData,
    selectedKeys: treeCheckable ? [] : [...valueSet],
    nodeRenderer,
    onSelect: (keys, _action) => {
      const action: SelectAction = {
        type: _action.selected ? SelectActionType.Select : SelectActionType.DeSelect,
        payload: _action.key,
      };
      set(keys, { action });
    },
    // ==================== Filter Tree Node ====================
    filterTreeNode: useCallback(
      node => {
        if (filterTreeNodeProp === false) {
          return true;
        } else if (isFunction(filterTreeNodeProp)) {
          return filterTreeNodeProp(inputValue, node);
        }
        return defaultFilterTreeNode(inputValue, node);
      },
      [filterTreeNodeProp, inputValue],
    ),
    showCheckedStrategy,
    switcherIcon,
    loadData,
    height: listHeight - 3 * 2 - 1 * 2, // 减去 padding border
    virtual,
    loadedKeys: treeLoadedKeys,
    onLoad: onTreeLoad,
  };

  if (!treeCheckStrictly) {
    treeProps = {
      ...treeProps,
      checkedKeys: treeCheckable ? [...valueSet] : [],
      checkStrictly: false,
      onCheck: (keys, _action) => {
        const action: SelectAction = {
          type: _action.checked ? SelectActionType.Checked : SelectActionType.UnChecked,
          payload: _action.key,
        };
        set(keys, { closeOverlay: false, action });
      },
    };
  } else {
    treeProps = {
      ...treeProps,
      defaultCheckedKeys: { checked: [], halfChecked: [] },
      checkedKeys: treeCheckable ? { checked: [...valueSet], halfChecked: [] } : { checked: [], halfChecked: [] },
      checkStrictly: true,
      onCheck: (keys, _action) => {
        const action: SelectAction = {
          type: _action.checked ? SelectActionType.Checked : SelectActionType.UnChecked,
          payload: _action.key,
        };
        set(keys.checked, { closeOverlay: false, action });
      },
    };
  }

  const treeCtxValue = useTree(treeProps, treeRef);

  const menuNode = (
    <TreeContextProvider value={treeCtxValue}>
      <TreeContainer />
    </TreeContextProvider>
  );

  return (
    <SelectOverlayContainer
      className={cssScope('__overlay')}
      optionsLength={treeData?.length ?? 0}
      filteredOptionsLength={treeCtxValue.filteredTreeNodes.size}
      menu={menuNode}
    />
  );
};

/**
 * @docgen
 */
export type TreeSelectRef = RefSelectRef & Pick<TreeRef, 'scrollIntoView' | 'toggleExpanded'>;

const _TreeSelect = React.forwardRef(function TreeSelectInner(
  props: PropsWithChildren<TreeSelectProps>,
  ref: Ref<TreeSelectRef>,
) {
  const {
    treeData: treeDataProp,
    treeCheckable = false,
    treeDefaultExpandAll = false,
    treeDefaultExpandedKeys = [],
    treeExpandedKeys,
    onTreeExpand,
    listHeight = 232,
    filterTreeNode: filterTreeNodeProp = true,
    showCheckedStrategy = ShowCheckedStrategy.SHOW_CHILD,
    treeCheckStrictly = false,
    switcherIcon,
    loadData,
    children,
    virtual,
    nodeRenderer,
    treeLoadedKeys,
    onTreeLoad,
    ...baseSelectProps
  } = props;

  const selectRef = React.useRef<RefSelectRef>(null);
  const treeRef = React.useRef<TreeRef>(null);
  const treeData = useMemo(() => treeDataProp ?? treeChildren2TreeNodes(children), [children, treeDataProp]);

  const { doublyTreeNodeMap } = useMemo(() => genDoublyTreeNodeMap(treeData || []), [treeData]);

  const treeCheckCtx = useTreeCheck(
    {
      doublyTreeNodeMap,
      checkStrategy: strategyMap[showCheckedStrategy],
      ...(treeCheckStrictly
        ? {
            checkStrictly: true,
            defaultCheckedKeys: toCheckStrictlyValue(props.defaultValue),
            ...('value' in props ? { checkedKeys: toCheckStrictlyValue(props.value) } : {}),
          }
        : {
            checkStrictly: false,
            defaultCheckedKeys: genArray(props.defaultValue),
            ...('value' in props ? { checkedKeys: genArray(props.value) } : {}),
          }),
    },

    { skipped: !treeCheckable },
  );
  const { checkedSet, setCheckedKeys } = treeCheckCtx;

  const checkedKeys = React.useMemo(() => Object.keys(checkedSet).filter(key => checkedSet[key]), [checkedSet]);

  /**
   * Select Props
   */
  let multiple: boolean = props.multiple ?? false;
  if (treeCheckable) {
    multiple = true;
  }
  baseSelectProps.multiple = multiple;

  const internalOptions = React.useMemo(() => {
    const options: InternalSelectProps<string, TreeNodeConfig>['options'] = [];
    traverseTreeConfig(treeData, node => {
      options.push(node);
    });
    return options;
  }, [treeData]);

  const overlayNode = (
    <TreeCheckedProvider value={treeCheckCtx}>
      <TreeSelectOverlay
        treeRef={treeRef}
        {...{
          treeData,
          treeCheckable,
          treeDefaultExpandAll,
          treeDefaultExpandedKeys,
          treeExpandedKeys,
          onTreeExpand,
          listHeight,
          multiple,
          filterTreeNode: filterTreeNodeProp,
          showCheckedStrategy,
          treeCheckStrictly,
          switcherIcon,
          loadData,
          virtual,
          nodeRenderer,
          treeLoadedKeys,
          onTreeLoad,
        }}
      />
    </TreeCheckedProvider>
  );

  let selectProps: InternalSelectProps<string, TreeNodeConfig>;

  if (baseSelectProps.multiple) {
    if (treeCheckable) {
      if ('value' in props) {
        // 受控模式
        baseSelectProps.value = checkedKeys;
      } else {
        // 非受控模式
        baseSelectProps.defaultValue = checkedKeys;
      }
    }

    const _onChange: MultipleInternalSelectProps<string, TreeNodeConfig>['onChange'] = (value, meta) => {
      if (treeCheckable) {
        const newValue = value;
        setCheckedKeys(treeCheckStrictly ? toCheckStrictlyValue(newValue) : genArray(newValue));
      }

      if (baseSelectProps.onChange) {
        const path = value.length
          ? value.reduce((a, b, index) => {
              const _path = getTreeNodeConfigPath(doublyTreeNodeMap[b]);
              a[index] = _path;
              return a;
            }, [] as TreeNodeConfig[][])
          : [];
        baseSelectProps.onChange(value, { ...meta, path });
      }
    };

    selectProps = {
      ...baseSelectProps,
      mode: 'multiple',
      options: internalOptions,
      overlayNode,
      onChange: _onChange,
    };
  } else {
    const _onChange: SingleInternalSelectProps<string, TreeNodeConfig>['onChange'] = (value, meta) => {
      if (baseSelectProps.onChange) {
        const path = getTreeNodeConfigPath(isNonNullable(value) ? doublyTreeNodeMap[value] : null);
        baseSelectProps.onChange(value, { ...meta, path });
      }
    };

    selectProps = {
      ...baseSelectProps,
      mode: 'single',
      options: internalOptions,
      overlayNode,
      onChange: _onChange,
    };
  }

  React.useImperativeHandle(ref, () => ({
    get input() {
      return selectRef.current?.input;
    },
    blur() {
      selectRef.current?.blur();
    },
    focus() {
      selectRef.current?.focus();
    },
    scrollIntoView(...args) {
      treeRef.current?.scrollIntoView(...args);
    },
    toggleExpanded(key) {
      treeRef.current?.toggleExpanded(key);
    },
  }));

  return <RefSelect ref={selectRef} {...selectProps} />;
}) as {
  (props: PropsWithChildren<MultipleTreeSelectProps> & RefAttributes<TreeSelectRef>): ReactElement | null;
  (
    // eslint-disable-next-line @typescript-eslint/unified-signatures
    props: PropsWithChildren<SingleTreeSelectProps> & RefAttributes<TreeSelectRef>,
  ): ReactElement | null;
};

export const TreeSelect = Object.assign(_TreeSelect, {
  SHOW_ALL: ShowCheckedStrategy.SHOW_ALL,
  SHOW_PARENT: ShowCheckedStrategy.SHOW_PARENT,
  SHOW_CHILD: ShowCheckedStrategy.SHOW_CHILD,
  TreeNode,
  [COMPONENT_VERSION_ID]: process.env.__VERSION__,
  displayName: 'TreeSelect',
});

export * from './interface';
