// @universe-design/react/src/tree-select/interface.ts
import type {
  BaseInternalSelectProps,
  SingleInternalSelectProps as _SingleInternalSelectProps,
  MultipleInternalSelectProps as _MultipleInternalSelectProps,
} from '../select/internal.interface';
import type { DoublyTreeNode, TreeProps } from '../tree';
import type { TreeNodeConfig } from '../tree/interface';

export interface TreeSelectFilterTreeNodeFunc {
  (inputValue: string, treeNode: DoublyTreeNode): boolean;
}

export interface SingleInternalTreeSelectProps
  extends Omit<_SingleInternalSelectProps<string, TreeNodeConfig>, 'mode'> {
  /**
   * 是否为多选
   */
  multiple?: false;
  treeCheckable?: false;
}

export interface MultipleInternalTreeSelectProps
  extends Omit<_MultipleInternalSelectProps<string, TreeNodeConfig>, 'mode'> {
  /**
   * 是否为多选
   */
  multiple: true;
  /**
   * 仅在多选模式下生效
   */
  treeCheckable?: boolean;
}

export interface TreeSelectOverlayProps
  extends Pick<TreeProps, 'treeData' | 'switcherIcon' | 'loadData' | 'virtual' | 'nodeRenderer'> {
  /**
   * 筛选 Tree 节点的筛选函数
   * @defaultValue true
   */
  filterTreeNode: TreeSelectFilterTreeNodeFunc | boolean;
  /**
   * Dropdown 的高度
   * @defaultValue 256
   */
  listHeight: number;
  /**
   * Tree 是否默认打开所有的父级
   * @defaultValue false
   */
  treeDefaultExpandAll: NonNullable<TreeProps['defaultExpandAll']>;
  /**
   * Tree 默认展开的 keys
   */
  treeDefaultExpandedKeys: NonNullable<TreeProps['defaultExpandedKeys']>;
  /**
   * Tree 展开的 keys，受控值
   */
  treeExpandedKeys?: TreeProps['expandedKeys'];
  /**
   * Tree 展开状态变化的回调
   */
  onTreeExpand?: TreeProps['onExpand'];
  /**
   * 定义选中项回填的方式。`TreeSelect.SHOW_ALL`： 显示所有选中节点，包括父节点。`TreeSelect.SHOW_PARENT`： 当父节点下所有子节点都选中时，只显示父节点。 `TreeSelect.SHOW_CHILD`： 当父节点下所有子节点都选中时，只显示子节点。默认显示所有节点
   * @defaultValue TreeSelect.SHOW_CHILD
   */
  showCheckedStrategy: NonNullable<TreeProps['showCheckedStrategy']>;
  /**
   * Checkable 状态下节点选择完全受控，父子节点选中状态不再关联
   * @defaultValue false
   */
  treeCheckStrictly: NonNullable<TreeProps['checkStrictly']>;
  /**
   * (tree) 已经加载的节点的受控值，需要配合 `loadData` 使用
   */
  treeLoadedKeys?: TreeProps['loadedKeys'];
  /**
   * (tree) 节点加载完毕时触发，你需要设置 `loadedKeys` 防止死循环
   */
  onTreeLoad?: TreeProps['onLoad'];
}

type OriginSingleOnChangeParams = Parameters<NonNullable<SingleInternalTreeSelectProps['onChange']>>;
type OriginMultipleOnChangeParams = Parameters<NonNullable<MultipleInternalTreeSelectProps['onChange']>>;

type SingleOnChangeType = (
  value: OriginSingleOnChangeParams[0],
  meta: OriginSingleOnChangeParams[1] & { path: TreeNodeConfig[] },
) => void;

type MultipleOnChangeType = (
  value: OriginMultipleOnChangeParams[0],
  meta: OriginMultipleOnChangeParams[1] & {
    path: TreeNodeConfig[][];
  },
) => void;

export type SingleTreeSelectProps = (Omit<SingleInternalTreeSelectProps, 'onChange'> & {
  onChange?: SingleOnChangeType;
}) &
  BaseInternalSelectProps<string, TreeNodeConfig> &
  Partial<TreeSelectOverlayProps>;

export type MultipleTreeSelectProps = (Omit<MultipleInternalTreeSelectProps, 'onChange'> & {
  onChange?: MultipleOnChangeType;
}) &
  BaseInternalSelectProps<string, TreeNodeConfig> &
  Partial<TreeSelectOverlayProps>;

/**
 * @public
 * @docgen
 */
export type TreeSelectProps = (
  | (Omit<SingleInternalTreeSelectProps, 'onChange'> & {
      onChange?: SingleOnChangeType;
    })
  | (Omit<MultipleInternalTreeSelectProps, 'onChange'> & {
      onChange?: MultipleOnChangeType;
    })
) &
  BaseInternalSelectProps<string, TreeNodeConfig> &
  Partial<TreeSelectOverlayProps>;
