// @universe-design/react/lib/badge/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  BaseBadge: true,
  Badge: true
};
exports.BaseBadge = exports.Badge = void 0;
var _uko6l5r = _interopRequireDefault(require("./../.stylex/uko6l5r"));
var _u5jxv = _interopRequireDefault(require("./../.stylex/u5jxv71"));
var _u1u1p7mj = _interopRequireDefault(require("./../.stylex/u1u1p7mj"));
var _u1r95smd = _interopRequireDefault(require("./../.stylex/u1r95smd"));
var _ulb4spa = _interopRequireDefault(require("./../.stylex/ulb4spa"));
var _u16uuw9n = _interopRequireDefault(require("./../.stylex/u16uuw9n"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _MoreOutlined = _interopRequireDefault(require("@universe-design/icons-react/MoreOutlined"));
var _clsx = _interopRequireDefault(require("clsx"));
var _react = _interopRequireDefault(require("react"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _stylex2 = require("../stylex");
var _util = require("../util");
var _composeRefs = _interopRequireDefault(require("../util/composeRefs"));
var _reactUtilities = require("@universe-design/react-utilities");
var _utils = require("./utils");
require("./style");
var _SVGTextBadge = require("./SVGTextBadge");
Object.keys(_SVGTextBadge).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _SVGTextBadge[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _SVGTextBadge[key];
    }
  });
});
var _CanvasTextBadge = require("./CanvasTextBadge");
Object.keys(_CanvasTextBadge).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _CanvasTextBadge[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _CanvasTextBadge[key];
    }
  });
});
var _types = require("./types");
Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _types[key];
    }
  });
});
//.u1xvp9sr{transform:none;}
//.u10pi13n{position:relative;}
//.u1euv43f{position:absolute;}
//.u1wirg6b{z-index:auto;}
//.uqzzkal{font-family:"PingFang SC","Microsoft YaHei","Myriad Pro","Hiragino Sans GB",sans-serif;}
//.unmn6fi{font-size:10px;}
/* eslint-disable max-lines-per-function */
/* eslint-disable @typescript-eslint/indent */
const MAX_BADGE_DISPLAY_NUMBER = 999;
const BADGE_FONT_SIZE = 10;
const cssBadgeBadge = _stylex.default.compose(_stylex2.cssBlockBase, _stylex2.cssFontMedium, {
  "props": {
    "qhf8xq": _u1u1p7mj.default,
    "Bj3rh1h": _u1r95smd.default,
    "Bahqtrf": _ulb4spa.default,
    "Be2twd7": _u16uuw9n.default
  }
});
const BaseBadge = exports.BaseBadge = /*#__PURE__*/_react.default.forwardRef((props, ref) => {
  const {
    children,
    status,
    offset,
    style: styleProp,
    className,
    title
  } = props;
  let {
    count
  } = props;
  const {
    badgeCssScope,
    shape,
    position,
    isHidden,
    showAsDot,
    hasStatus
  } = (0, _utils.useBadge)(props);

  // Title
  const titleNode = title !== null && title !== void 0 ? title : typeof count === 'string' || typeof count === 'number' ? String(count) : undefined;
  if (isHidden) {
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, children);
  }
  const hasChild = Boolean(children);
  let style = {
    ...styleProp
  };
  let colorClassName = '';
  const badgeBadgeCls = (0, _clsx.default)(badgeCssScope('__badge'), !children && badgeCssScope('-has-no-child'), (0, _stylex.default)(cssBadgeBadge, !children && {
    "props": {
      "Bz10aip": _uko6l5r.default,
      "qhf8xq": _u5jxv.default
    }
  }));
  if (hasChild) {
    /**
     * 📢 定位要用 translate3d，在某些场景下可能出现 badge 内容左右偏移，加这个可以让 badge 独立一层
     */
    if (position === 'topRight') {
      style = {
        top: 0,
        right: 0,
        transform: offset ? `translate3d(calc(1 * (${(0, _utils.normalizeCssSize)(offset[0])})), calc(-1 * (${(0, _utils.normalizeCssSize)(offset[1])})), 0)` : undefined,
        ...style
      };
    } else if (position === 'topLeft') {
      style = {
        top: 0,
        left: 0,
        transform: offset ? `translate3d(calc(-1 * (${(0, _utils.normalizeCssSize)(offset[0])})), calc(-1 * (${(0, _utils.normalizeCssSize)(offset[1])})), 0)` : undefined,
        ...style
      };
    } else if (position === 'bottomRight') {
      style = {
        bottom: 0,
        right: 0,
        transform: offset ? `translate3d(calc(1 * (${(0, _utils.normalizeCssSize)(offset[0])})), calc(1 * (${(0, _utils.normalizeCssSize)(offset[1])})), 0)` : undefined,
        ...style
      };
    } else {
      style = {
        bottom: 0,
        left: 0,
        transform: offset ? `translate3d(calc(-1 * (${(0, _utils.normalizeCssSize)(offset[0])})), calc(1 * (${(0, _utils.normalizeCssSize)(offset[1])})), 0)` : undefined,
        ...style
      };
    }
  }
  if (showAsDot) {
    // Dot badge, Status badge
    const {
      colorStyle = 'filled',
      color,
      text,
      size
    } = props;
    const radius = (0, _utils.getBadgeRadius)({
      size,
      dot: true
    });
    style = {
      width: radius * 2,
      height: radius * 2,
      flexShrink: 0,
      ...style
    };
    if (color) {
      // color 优先级高于 status
      if ((0, _utils.isPresetColor)(color)) {
        colorClassName = (0, _clsx.default)(badgeCssScope(`-${color}-${colorStyle}`), badgeCssScope(`-${colorStyle}`));
      } else if (colorStyle === 'filled') {
        style = {
          backgroundColor: color,
          ...style
        };
      } else {
        style = {
          backgroundColor: 'transparent',
          border: `1px solid ${color}`,
          ...style
        };
      }
    } else if (status) {
      colorClassName = (0, _clsx.default)(badgeCssScope(`-${status}-${colorStyle}`), badgeCssScope(`-${colorStyle}`));
    } else {
      colorClassName = (0, _clsx.default)(badgeCssScope(`-red-${colorStyle}`), badgeCssScope(`-${colorStyle}`));
    }
    if (!hasStatus) {
      return /*#__PURE__*/_react.default.createElement("span", {
        className: (0, _clsx.default)(badgeCssScope(), badgeCssScope('-dot'), className)
      }, children, /*#__PURE__*/_react.default.createElement("sup", {
        ref: ref,
        title: titleNode,
        className: (0, _clsx.default)(badgeBadgeCls, badgeCssScope(`-${position}`), colorClassName),
        style: style
      }));
    } else {
      return /*#__PURE__*/_react.default.createElement("span", {
        className: (0, _clsx.default)(badgeCssScope(), badgeCssScope('-dot-inline'), className)
      }, /*#__PURE__*/_react.default.createElement("span", {
        ref: ref,
        className: (0, _clsx.default)(badgeBadgeCls, colorClassName),
        style: style
      }), (0, _util.isNonNullable)(text) && /*#__PURE__*/_react.default.createElement("span", {
        className: (0, _clsx.default)(badgeCssScope('__text'))
      }, text));
    }
  } else {
    // Text badge
    let type = 'text';
    const {
      overflowCount = MAX_BADGE_DISPLAY_NUMBER,
      color = 'red',
      hack10px = false,
      textBadgeSidecar: TextBadgeSidecar
    } = props;

    // 处理 Sidecar 逻辑
    const supClassName = (0, _clsx.default)(badgeBadgeCls, badgeCssScope(`-${position}`));
    const isCountOverflow = typeof count === 'number' && count > overflowCount;
    if (TextBadgeSidecar && typeof count !== 'object' && !isCountOverflow) {
      return /*#__PURE__*/_react.default.createElement("span", {
        className: (0, _clsx.default)(badgeCssScope(), className)
      }, children, /*#__PURE__*/_react.default.createElement(TextBadgeSidecar, {
        color: color,
        className: supClassName,
        style: style,
        content: String(count)
      }));
    }
    if ((0, _utils.isPresetColor)(color)) {
      colorClassName = badgeCssScope(`-${color}-filled`);
    } else {
      style = {
        backgroundColor: color,
        ...style
      };
    }
    if (shape === 'circle' || String(count).length === 1) {
      type = 'icon';
    }
    if (isCountOverflow) {
      count = /*#__PURE__*/_react.default.createElement(_MoreOutlined.default, {
        className: badgeCssScope('__badge__more')
      });
    }
    return /*#__PURE__*/_react.default.createElement("span", {
      className: (0, _clsx.default)(badgeCssScope(), badgeCssScope(`-${type}`), className)
    }, children, /*#__PURE__*/_react.default.createElement("sup", {
      ref: ref,
      title: titleNode,
      className: (0, _clsx.default)(supClassName, colorClassName),
      style: style
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: (0, _clsx.default)(badgeCssScope('__badge__content'), hack10px && badgeCssScope('__badge__content-hack-10px'))
    }, count)));
  }
});
const Badge = exports.Badge = /*#__PURE__*/_react.default.forwardRef((props, ref) => {
  const {
    children,
    offset: offsetProp,
    position,
    contentType,
    childCornerRadius: childCornerRadiusProp,
    ...restProps
  } = props;
  const {
    showAsDot
  } = (0, _utils.useBadge)(props);
  const shouldGetChildrenBorderRadius = !(0, _util.isNonNullable)(childCornerRadiusProp) && !(0, _util.isNonNullable)(offsetProp);
  const [childCornerRadius, setChildCornerRadius] = (0, _reactUtilities.useControlledState)(0, childCornerRadiusProp);
  const childRef = (0, _reactUtilities.useProxyRef)(null, node => {
    if (node && node instanceof HTMLElement && shouldGetChildrenBorderRadius) {
      setChildCornerRadius((0, _utils.getBorderRadius)(node, position));
    }
  });
  const mergedChildRef = _react.default.useMemo(() => (0, _composeRefs.default)(childRef, children === null || children === void 0 ? void 0 : children.ref), [childRef, children]);
  const badgeRadius = (0, _utils.getBadgeRadius)(props);
  const badgeCenterPoint = {
    x: badgeRadius,
    y: badgeRadius
  };
  const tangencyPoint = {
    x: (1 - Math.SQRT1_2) * childCornerRadius,
    y: (1 - Math.SQRT1_2) * childCornerRadius
  };
  // Offset
  const offset = _react.default.useMemo(() => {
    let _offset = [-tangencyPoint.x + badgeCenterPoint.x, -tangencyPoint.y + badgeCenterPoint.y];
    if (contentType === 'text' && showAsDot) {
      _offset = ['2px + 100%', 0];
    } else if (contentType === 'text') {
      _offset = ['2px + 100%', '50%'];
    } else if ((0, _util.isNonNullable)(offsetProp)) {
      _offset = offsetProp;
    }
    return _offset;
  }, [tangencyPoint.x, tangencyPoint.y, badgeCenterPoint.x, badgeCenterPoint.y, contentType, showAsDot, offsetProp]);

  // 如果是有 offsetProp 或者 children 没法 cloneELement 或者不需要获取子元素圆角的时候直接返回 base badge
  if (! /*#__PURE__*/_react.default.isValidElement(children) || !shouldGetChildrenBorderRadius) {
    return /*#__PURE__*/_react.default.createElement(BaseBadge, (0, _extends2.default)({}, restProps, {
      ref: ref,
      offset: offset,
      position: position
    }), children);
  }
  return /*#__PURE__*/_react.default.createElement(BaseBadge, (0, _extends2.default)({}, restProps, {
    ref: ref,
    offset: offset,
    position: position
  }), /*#__PURE__*/_react.default.cloneElement(children, {
    ref: mergedChildRef
  }));
});
