// @universe-design/react/lib/badge/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getBadgeRadius = exports.getBadgeFont = exports.defaults = void 0;
exports.getBorderRadius = getBorderRadius;
exports.measureText = exports.isPresetColor = void 0;
exports.normalizeCssSize = normalizeCssSize;
exports.useBadge = useBadge;
var _reactUtilities = require("@universe-design/react-utilities");
var _hooks = require("../config-provider/hooks");
var _util = require("../util");
const defaults = exports.defaults = {
  textBadgeFontSize: 15
};
const EMPTY_TEXT_METRICS = {
  actualBoundingBoxAscent: 0,
  actualBoundingBoxDescent: 0,
  actualBoundingBoxLeft: 0,
  actualBoundingBoxRight: 0,
  width: 0,
  fontBoundingBoxAscent: 0,
  fontBoundingBoxDescent: 0,
  alphabeticBaseline: 0,
  emHeightAscent: 0,
  emHeightDescent: 0,
  hangingBaseline: 0,
  ideographicBaseline: 0
};
let ctxForMeasure = null;
const measureText = (content, font) => {
  if (!content.length || !(0, _reactUtilities.isClient)()) {
    return EMPTY_TEXT_METRICS;
  }
  if (!ctxForMeasure) {
    ctxForMeasure = document.createElement('canvas').getContext('2d');
  }
  if (!ctxForMeasure) {
    return EMPTY_TEXT_METRICS;
  }
  ctxForMeasure.font = font;
  const dim = ctxForMeasure.measureText(content);
  return dim;
};
exports.measureText = measureText;
const PresetColorRegex = new RegExp(`^(${_util.PresetColorEnum.join('|')})$`);
const isPresetColor = color => PresetColorRegex.test(color);
exports.isPresetColor = isPresetColor;
const getBadgeFont = fontSize => `500 ${fontSize}px "PingFang SC", "Microsoft YaHei", "Myriad Pro", "Hiragino Sans GB", sans-serif`;

/**
 * 根据 prop 计算 badge 的半径
 */
exports.getBadgeFont = getBadgeFont;
const getBadgeRadius = _ref => {
  let {
    size = 'medium',
    dot = false
  } = _ref;
  if (!dot) {
    return 7;
  }
  let radius = 4;
  if (typeof size === 'number') {
    radius = size / 2;
  } else if (size === 'large') {
    radius = 5;
  } else if (size === 'small') {
    radius = 3;
  }
  return radius;
};

/**
 * 获取方形或圆形头像的 border radius px 值
 * 不适用于通用场景
 */
exports.getBadgeRadius = getBadgeRadius;
function getBorderRadius(el, position) {
  const css = getComputedStyle(el);

  // webit-border-radius 为兼容 firefox
  let borderRadiusString = css.borderTopRightRadius || css.webkitBorderTopRightRadius;
  if (position === 'bottomLeft') {
    borderRadiusString = css.borderBottomLeftRadius || css.webkitBorderBottomLeftRadius;
  } else if (position === 'bottomRight') {
    borderRadiusString = css.borderBottomRightRadius || css.webkitBorderBottomRightRadius;
  } else if (position === 'topLeft') {
    borderRadiusString = css.borderTopLeftRadius || css.webkitBorderTopLeftRadius;
  }
  const borderRadius = parseFloat(borderRadiusString);
  if (borderRadiusString.match(/^\d+(?:\.\d+)?%$/)) {
    return Math.min(el.offsetHeight * borderRadius / 100, el.offsetHeight / 2);
  }
  if (Object.is(borderRadius, NaN)) {
    return 0;
  }
  return Math.min(borderRadius, el.offsetHeight / 2);
}
function useBadge(props) {
  const {
    status,
    position = 'topRight',
    showZero = false,
    dot = false,
    shape = 'round',
    count
  } = props;
  const badgeCssScope = (0, _hooks.useCssScope)('badge');
  // ======== 如果不 visible 直接返回 children  ========
  const hasStatus = (0, _util.isNonNullable)(status) || props.text;
  const isZero = count === '0' || count === 0;
  /**
   * 显示为点状徽标还是文字徽标
   */
  const showAsDot = dot && !isZero || hasStatus;
  const isEmpty = !(0, _util.isNonNullable)(count) || count === '';
  /**
   * 是否隐藏徽标
   */

  return {
    position,
    shape,
    badgeCssScope,
    hasStatus,
    showAsDot,
    isHidden: (isEmpty || isZero && !showZero) && !showAsDot
  };
}
function normalizeCssSize(s) {
  return typeof s === 'number' ? `${s}px` : s;
}
