// @byted-nodex/metrics/lib/reservoir/addition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Addition = void 0;
const reservoir_1 = require("./reservoir");
class Addition extends reservoir_1.Reservoir {
    get active() {
        // 冗余一次，golang 是 3 次，node 设置保守一些
        return this.idle <= (this.policy.reportUnchangeCounter ? 1 : 0);
    }
    realMerge(val) {
        // counter 累加
        this.value.value += val.value;
        return true;
    }
    values() {
        const vs = [
            {
                suffix: this.value.suffix,
                type: this.value.type,
                values: [this.value.value],
                lastValues: this.setLastValue ? [this.lastValue] : undefined,
            },
        ];
        this.lastValue = this.value.value;
        this.setLastValue = false;
        return vs;
    }
    skip() {
        this.setLastValue = this.policy.reportLastValue;
    }
    init() {
        this.setLastValue = this.policy.reportInitialValue;
        this.lastValue = 0;
        this.value.value = 0;
    }
    reset() {
        // Counter 应该一直累加，所以无需 reset
    }
}
exports.Addition = Addition;
