// @byted-nodex/metrics/lib/reservoir/histogram.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Histogram = void 0;
const reservoir_1 = require("./reservoir");
class Histogram extends reservoir_1.Reservoir {
    realMerge(val) {
        const v = val.value;
        for (let i = 0; i < this.buckets.length; i++) {
            const bucket = this.buckets[i];
            let right;
            if (i < this.buckets.length - 1) {
                right = this.buckets[i + 1].left;
            }
            else {
                right = Infinity;
            }
            if (v >= bucket.left && v < right) {
                bucket.count++;
                this.sum += v;
                this.count++;
                break;
            }
        }
        return true;
    }
    values() {
        if (this.count === 0) {
            return null;
        }
        const vs = [];
        this.buckets.forEach(bucket => {
            vs.push(bucket.count);
            bucket.count = 0;
        });
        vs.push(this.sum, this.count);
        return [
            {
                suffix: this.value.suffix,
                type: this.value.type,
                values: vs,
            },
        ];
    }
    init() {
        this.reset();
    }
    reset() {
        var _a;
        this.sum = 0;
        this.count = 0;
        this.buckets =
            ((_a = this.value.buckets) === null || _a === void 0 ? void 0 : _a.map(b => ({
                left: b,
                count: 0,
            }))) || [];
    }
}
exports.Histogram = Histogram;
