// @byted-nodex/metrics/lib/reservoir/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Reservoir = exports.Queue = void 0;
const metrics_codec_1 = require("@byted-nodex/metrics-codec");
const reservoir_1 = require("./reservoir");
Object.defineProperty(exports, "Reservoir", { enumerable: true, get: function () { return reservoir_1.Reservoir; } });
const addition_1 = require("./addition");
const ratio_1 = require("./ratio");
const idempotence_1 = require("./idempotence");
const summary_1 = require("./summary");
const histogram_1 = require("./histogram");
const delta_counter_1 = require("./delta_counter");
const util_1 = require("../util");
function newReservoir(val) {
    let res;
    switch (val.type) {
        case metrics_codec_1.MetricType.Counter:
            res = new addition_1.Addition(val);
            break;
        case metrics_codec_1.MetricType.RateCounter:
            res = new ratio_1.Ratio(val);
            break;
        case metrics_codec_1.MetricType.Store:
            res = new idempotence_1.Idempotence(val);
            break;
        case metrics_codec_1.MetricType.Timer:
            res = new summary_1.Summary(val);
            break;
        case metrics_codec_1.MetricType.Histogram:
            res = new histogram_1.Histogram(val);
            break;
        case metrics_codec_1.MetricType.DeltaCounterType:
            res = new delta_counter_1.DeltaCounter(val);
            break;
        default:
            throw new Error('Unknown type');
    }
    return res;
}
class Queue {
    constructor() {
        this.size = 0;
    }
    /** 计算指标 */
    marshall(cb) {
        let current = this.head;
        this.active = false;
        while (current) {
            if (current.active) {
                this.active = true;
                const vs = current.marshal();
                if (vs) {
                    cb(vs, current);
                }
            }
            else {
                current.skip();
            }
            current = current.next;
        }
    }
    /** 分析缓存数据，用于 dump 结果排查用 */
    analyse(info) {
        let current = this.head;
        while (current) {
            const detail = (info.detail[current.type] = info.detail[current.type] || {
                metricCount: 0,
                activeMetricCount: 0,
            });
            if (current.active) {
                info.activeMetricCount++;
                detail.activeMetricCount++;
            }
            info.metricCount++;
            detail.metricCount++;
            current = current.next;
        }
    }
    // 通过单向链表合并不同打点项
    mergeOrInsert(value) {
        if (value.suffix && !(0, util_1.isValidString)(value.suffix)) {
            throw new Error(`invalid metric name suffix: ${value.suffix}`);
        }
        this.active = true;
        if (!this.head) {
            this.head = newReservoir(value);
            this.size++;
            return;
        }
        let current = this.head;
        while (current) {
            if (current.merge(value)) {
                // 合并成功不继续遍历
                break;
            }
            if (!current.next) {
                // 没有可合并项，新增节点
                current.next = newReservoir(value);
                this.size++;
                break;
            }
            else {
                current = current.next;
            }
        }
    }
    gc({ expiredDuration, ignoreMetrics }) {
        const currTime = Date.now();
        let curr = this.head;
        let prev;
        while (curr) {
            if (!curr.active &&
                (!expiredDuration || currTime - curr.lastActiveTime > expiredDuration) &&
                (!ignoreMetrics || !ignoreMetrics.includes(curr.type))) {
                this.size--;
                // 节点过期，从链表中删除节点
                if (!prev) {
                    this.head = curr.next;
                }
                else {
                    prev.next = curr.next;
                    curr.next = undefined;
                    curr = prev.next;
                    continue;
                }
            }
            prev = curr;
            curr = curr.next;
        }
    }
}
exports.Queue = Queue;
