// @byted-nodex/metrics/lib/reservoir/reservoir.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Reservoir = void 0;
class Reservoir {
    get type() {
        return this.value.type;
    }
    get policy() {
        return this.value.policy;
    }
    get suffix() {
        return this.value.suffix;
    }
    get active() {
        return this.idle === 0;
    }
    constructor(val) {
        this.lastActiveTime = 0;
        this.value = { ...val };
        this.init();
        this.merge(val);
    }
    merge(val) {
        if (val.type !== this.value.type) {
            return false;
        }
        if (val.suffix !== this.value.suffix) {
            return false;
        }
        const mergeResult = this.realMerge(val);
        if (mergeResult) {
            this.lastActiveTime = Date.now();
            this.idle = 0;
        }
        return mergeResult;
    }
    marshal() {
        const values = this.values();
        this.reset();
        this.idle++;
        return values;
    }
    skip() {
        // do nothing
    }
}
exports.Reservoir = Reservoir;
