// @byted-nodex/metrics/lib/reservoir/summary.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Summary = void 0;
/**
 * Timer
 * 暂存采样周期内的打点，然后聚合计算 .min/.max/.avg/.sum/.counter/.pct50/.pct90/.pct95/.pct99/.pct999
 */
const metrics_codec_1 = require("@byted-nodex/metrics-codec");
const util_1 = require("../util");
const consts_1 = require("../consts");
const reservoir_1 = require("./reservoir");
class Summary extends reservoir_1.Reservoir {
    realMerge(val) {
        var _a, _b;
        const v = val.value;
        if (v < this.min) {
            this.min = v;
        }
        if (v > this.max) {
            this.max = v;
        }
        this.sum += v;
        this.count++;
        if (((_a = this.policy) === null || _a === void 0 ? void 0 : _a.timerCacheCapacity) && this.hot.length > ((_b = this.policy) === null || _b === void 0 ? void 0 : _b.timerCacheCapacity)) {
            return true;
        }
        this.hot.push(v);
        return true;
    }
    values() {
        var _a;
        if (!this.hot.length) {
            return null;
        }
        // 升序排序
        const hot = this.hot.sort((a, b) => a - b);
        // 顺序需要保证
        const results = [
            this.min, // min
            this.max, // max
            this.sum / this.count, // avg
            this.sum, // sum
            this.count, // counter
            (0, util_1.percentile)(hot, 0.5), // pct50
            (0, util_1.percentile)(hot, 0.9), // pct90
            (0, util_1.percentile)(hot, 0.95), // pct95
            (0, util_1.percentile)(hot, 0.99), // pct99
            (0, util_1.percentile)(hot, 0.999), // pct999
        ];
        let values;
        if ((_a = this.policy) === null || _a === void 0 ? void 0 : _a.compatTimer) {
            // 单值
            values = consts_1.defaultTimerFieldsSuffixes.map((suffix, index) => {
                return {
                    suffix: this.value.suffix ? `${this.value.suffix}.${suffix}` : suffix,
                    type: metrics_codec_1.MetricType.Store,
                    values: [results[index]],
                };
            });
        }
        else {
            // 多值
            values = [{ suffix: this.value.suffix, type: this.value.type, values: results }];
        }
        return values;
    }
    init() {
        this.reset();
    }
    reset() {
        this.hot = [];
        this.sum = 0;
        this.max = 0;
        this.min = Infinity;
        this.count = 0;
    }
}
exports.Summary = Summary;
