// @byted-service/logger/lib/streams/file_rotate/cron_expression.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CornExpression = void 0;
const tslib_1 = require("tslib");
const dayjs_1 = tslib_1.__importDefault(require("dayjs"));
class CornExpression {
    constructor(pattern) {
        const parts = pattern.split(' ');
        this.pattern = pattern;
        this.seconds = this.getRange(parts[0], 0, 59);
        this.secondToIndex = {};
        for (let i = 0; i < this.seconds.length; i++) {
            this.secondToIndex[this.seconds[i]] = i;
        }
        this.minutes = this.getRange(parts[1], 0, 59);
        this.minuteToIndex = {};
        for (let i = 0; i < this.minutes.length; i++) {
            this.minuteToIndex[this.minutes[i]] = i;
        }
        this.hours = this.getRange(parts[2], 0, 23);
        this.hourToIndex = {};
        for (let i = 0; i < this.hours.length; i++) {
            this.hourToIndex[this.hours[i]] = i;
        }
        this.dates = this.getRange(parts[3], 1, 31);
        this.dateToIndex = {};
        for (let i = 0; i < this.dates.length; i++) {
            this.dateToIndex[this.dates[i]] = i;
        }
        this.months = this.getRange(parts[4], 1, 12);
        this.monthToIndex = {};
        for (let i = 0; i < this.months.length; i++) {
            this.monthToIndex[this.months[i]] = i;
        }
    }
    prev(now) {
        const time = now || new Date();
        const nowSecond = time.getSeconds();
        let nowMinute = time.getMinutes();
        let nowHour = time.getHours();
        let nowDate = time.getDate();
        const nowMonth = time.getMonth() + 1;
        const nowYear = time.getFullYear();
        let second = this.prevNum(nowSecond, this.seconds, this.secondToIndex, 60);
        if (second < 0) {
            second += 60;
            nowMinute -= 1;
        }
        let minute = this.prevNum(nowMinute + 1, this.minutes, this.minuteToIndex, 60);
        if (minute < 0) {
            minute += 60;
            nowHour -= 1;
        }
        let hour = this.prevNum(nowHour + 1, this.hours, this.hourToIndex, 24);
        if (hour < 0) {
            hour += 24;
            nowDate -= 1;
        }
        const date = this.prevNum(nowDate + 1, this.dates, this.dateToIndex, 31);
        const month = this.prevNum(nowMonth + 1, this.months, this.monthToIndex, 12);
        const year = nowYear;
        return (0, dayjs_1.default)(new Date(year, month - 1, date, hour, minute, second));
    }
    next(now) {
        const time = now || new Date();
        const nowSecond = time.getSeconds();
        let nowMinute = time.getMinutes();
        let nowHour = time.getHours();
        let nowDate = time.getDate();
        let nowMonth = time.getMonth() + 1;
        let nowYear = time.getFullYear();
        let second = this.nextNum(nowSecond, this.seconds, this.secondToIndex, 60);
        if (second >= 60) {
            second -= 60;
            nowMinute += 1;
        }
        let minute = this.nextNum(nowMinute - 1, this.minutes, this.minuteToIndex, 60);
        if (minute >= 60) {
            minute -= 60;
            nowHour += 1;
        }
        let hour = this.nextNum(nowHour - 1, this.hours, this.hourToIndex, 24);
        if (hour >= 24) {
            hour -= 24;
            nowDate += 1;
        }
        let date = this.nextNum(nowDate - 1, this.dates, this.dateToIndex, 31);
        if (date > 31) {
            date -= 31;
            nowMonth += 1;
        }
        let month = this.nextNum(nowMonth - 1, this.months, this.monthToIndex, 12);
        if (month > 12) {
            month -= 12;
            nowYear += 1;
        }
        const year = nowYear;
        return (0, dayjs_1.default)(new Date(year, month - 1, date, hour, minute, second));
    }
    getRange(text, min, max) {
        // *
        if (text === '*') {
            const range = [];
            for (let i = min; i < max + 1; i++) {
                range.push(i);
            }
            return range;
        }
        // 0
        if (text.match(/^(-)?\d+$/)) {
            const num = Number(text);
            if (num < min || num > max) {
                throw new Error(`Expression is invalid "${this.pattern}". It must be >= ${min} and <= ${max}`);
            }
            return [num];
        }
        // 0-10
        if (text.match(/^\d+-\d+$/)) {
            const [minStr, maxStr] = text.split('-');
            const min_ = Number(minStr);
            const max_ = Number(maxStr);
            if (max_ < min_) {
                throw new Error(`Expression is invalid "${this.pattern}"`);
            }
            if (min_ < min || max_ > max) {
                throw new Error(`Expression is invalid "${this.pattern}". It must be >= ${min} and <= ${max}`);
            }
            const range = [];
            for (let i = min_; i < max_ + 1; i++) {
                range.push(i);
            }
            return range;
        }
        // 0,1,2
        if (text.match(/^(\d+,)+\d+$/)) {
            const nums = text
                .split(',')
                .map((n) => Number(n))
                .sort();
            const range = [];
            for (let i = 0; i < nums.length; i++) {
                const num = nums[i];
                if (num < min || num > max) {
                    throw new Error(`Expression is invalid "${this.pattern}". It must be >= ${min} and <= ${max}`);
                }
                range.push(num);
            }
            return range;
        }
        // */2
        if (text.match(/^\*\/\d+$/)) {
            const num = Number(text.slice(2));
            if (num < min || num > max) {
                throw new Error(`Expression is invalid "${this.pattern}". It must be >= ${min} and <= ${max}`);
            }
            if (num === 0) {
                throw new Error(`Expression is invalid "${this.pattern}"`);
            }
            const range = [];
            for (let i = min; i < max + 1; i++) {
                if (Number.isInteger(i / num)) {
                    range.push(i);
                }
            }
            return range;
        }
        throw new Error(`Unsupport expression "${this.pattern}"`);
    }
    prevNum(num, nums, numToIndex, mark) {
        const index = numToIndex[num];
        if (typeof index === 'number') {
            if (index === 0) {
                return nums[nums.length - 1] - mark;
            }
            else {
                return nums[index - 1];
            }
        }
        else {
            if (num <= nums[0]) {
                return nums[nums.length - 1] - mark;
            }
            else if (num >= nums[nums.length - 1]) {
                return nums[nums.length - 1];
            }
            else {
                let i = 1;
                while (nums[i] < num) {
                    i++;
                }
                return nums[i - 1];
            }
        }
    }
    nextNum(num, nums, numToIndex, mark) {
        const index = numToIndex[num];
        if (typeof index === 'number') {
            if (index === nums.length - 1) {
                return nums[0] + mark;
            }
            else {
                return nums[index + 1];
            }
        }
        else {
            if (num <= nums[0]) {
                return nums[0];
            }
            else if (num >= nums[nums.length - 1]) {
                return nums[0] + mark;
            }
            else {
                let i = 1;
                while (nums[i] <= num) {
                    i++;
                }
                return nums[i];
            }
        }
    }
}
exports.CornExpression = CornExpression;
