// @universe-design/react/es/table/hooks/useSelection/SelectionTitle.js
import React, { useState } from 'react';
import { Dropdown } from '../../../dropdown';
import { ExpandButton } from '../../../shared/components/expand-button';
import { Menu } from '../../../menu';
import { Checkbox } from '../../../checkbox';
function SelectionTitle(props) {
  const {
    prefixCls,
    checked,
    disabled,
    indeterminate,
    getPopupContainer,
    recordKeys,
    onSelectAllChange,
    mergedSelections
  } = props;
  let customizeSelections;
  if (mergedSelections) {
    const menu = /*#__PURE__*/React.createElement(Menu, {
      getPopupContainer: getPopupContainer
    }, mergedSelections.map((selection, index) => {
      const {
        key,
        text,
        onSelect
      } = selection;
      return /*#__PURE__*/React.createElement(Menu.Item, {
        key: key || index,
        onClick: () => {
          if (onSelect) {
            onSelect(recordKeys);
          }
        }
      }, text);
    }));

    // eslint-disable-next-line react-hooks/rules-of-hooks
    const [visible, setVisible] = useState(false);
    customizeSelections = /*#__PURE__*/React.createElement(Dropdown, {
      visible: visible,
      onVisibleChange: e => setVisible(e),
      overlay: menu,
      getPopupContainer: getPopupContainer
    }, /*#__PURE__*/React.createElement(ExpandButton, {
      isExpanded: visible,
      className: `${prefixCls}-selection-extra`,
      size: "xxs",
      iconType: "DownBoldOutlined"
    }));
  }
  return /*#__PURE__*/React.createElement("div", {
    className: `${prefixCls}-selection`
  }, /*#__PURE__*/React.createElement(Checkbox, {
    checked: checked,
    indeterminate: indeterminate,
    onChange: onSelectAllChange,
    disabled: disabled
  }), customizeSelections);
}
export default SelectionTitle;
