// @universe-design/react/src/anchor/AnchorLink.tsx
import React from 'react';
import clsx from 'clsx';
import stylex from '@universe-design/stylex';
import { cssTypographyHeadline, cssTypographyBody0 } from '../stylex';
import { isFunction } from '../util';
import { useCssScope } from '../config-provider/hooks';
import { useIsomorphicLayoutEffect as useLayoutEffect } from '../hooks';
import { useAnchorContext } from './AnchorContext';

import './style';

/**
 * Anchor.Link 组件的属性
 * @public
 */
export interface AnchorLinkProps extends Omit<React.HTMLAttributes<HTMLDivElement>, 'title' | 'onClick'> {
  /**
   * 锚点链接
   */
  href: string;
  /**
   * 指定在何处显示链接的资源
   */
  target?: string;
  /**
   * 文字内容
   */
  title: React.ReactNode;
  /**
   * 子元素
   */
  children?: React.ReactNode;
  /**
   * 点击事件
   */
  onClick?: (e: React.MouseEvent<HTMLElement>, link: { title: React.ReactNode; href: string }) => void;
}

export const AnchorLink = React.forwardRef<HTMLDivElement, AnchorLinkProps>((props, ref) => {
  const { href, title, target, children, className, onClick, ...restProps } = props;

  const cssScope = useCssScope('anchor-link');
  const { activeLink, registerLink, unregisterLink, handleLinkClick } = useAnchorContext();

  const linkRef = React.useRef<HTMLAnchorElement>(null);

  const cls = clsx(cssScope(''), className, {
    [cssScope('--active')]: activeLink === href,
  });

  const titleCls = clsx(
    cssScope('__title'),
    {
      [cssScope('__title--active')]: activeLink === href,
    },
    stylex(cssTypographyBody0, activeLink === href && cssTypographyHeadline),
  );

  useLayoutEffect(() => {
    registerLink(href, {
      ref: linkRef,
    });
    return () => unregisterLink(href);
  }, [href, registerLink, unregisterLink]);

  return (
    <div className={cls} ref={ref} {...restProps}>
      <a
        ref={linkRef}
        className={titleCls}
        href={href}
        title={typeof title === 'string' ? title : ''}
        target={target}
        onClick={e => {
          isFunction(onClick) && onClick(e, { title, href });
          handleLinkClick(e, { title, href });
        }}
      >
        {title}
      </a>
      {children}
    </div>
  );
});

if (__DEV__) {
  AnchorLink.displayName = 'AnchorLink';
}
