// @universe-design/react/src/table/hooks/useSelection/index.tsx
import type { ChangeEvent } from 'react';
import React, { useCallback, useMemo, useEffect } from 'react';
import { INTERNAL_COL_DEFINE } from '@universe-design/rc-table';
import difference from 'lodash/difference';
import without from 'lodash/without';
import { useControlledState } from '@universe-design/react-utilities';
import type { CheckboxProps } from '../../../checkbox/Checkbox';
import SelectionTitle from './SelectionTitle';
import SelectionCell from './SelectionCell';
import type {
  TableRowSelection,
  Key,
  ColumnsType,
  GetRowKey,
  TransformColumns,
  ExpandType,
  TableLocale,
  GetPopupContainerInTable,
  SelectionItem,
  // GetCheckDisabled
} from '../../interface';
import { SELECTION_TYPE_RADIO, SELECTION_TYPE_CHECKBOX, SELECTION_ALL, SELECTION_INVERT } from '../../constants';
import { isFunction } from '../../../util';
import {
  convertDataToEntities,
  checkedSet2Cache,
  CheckedType,
  unCheck,
  check,
  cache2CheckedSet,
  selectWithShift,
} from '../../utils';
import { cloneMap, array2Set } from '../../../tree';

export type INTERNAL_SELECTION_ITEM = SelectionItem | typeof SELECTION_ALL | typeof SELECTION_INVERT;

export interface UseSelectionConfig<RecordType> {
  prefix: string;
  pageData: readonly RecordType[];
  data: readonly RecordType[];
  getRowKey: GetRowKey<RecordType>;
  getRowChildren?: (date: RecordType) => RecordType[] | undefined;
  getRecordByKey: (key: Key) => RecordType;
  expandType: ExpandType;
  childrenColumnName: string;
  tableLocale: TableLocale;
  getPopupContainer?: GetPopupContainerInTable;
}

function flattenData<RecordType>(data: readonly RecordType[] | undefined, childrenColumnName: string): RecordType[] {
  let list: RecordType[] = [];
  if (!data) {
    return list;
  }

  data.forEach((record: RecordType) => {
    list.push(record);
    if (record && typeof record === 'object' && childrenColumnName in record) {
      const childrenData = (record as any)[childrenColumnName];
      list = [...list, ...flattenData<RecordType>(childrenData, childrenColumnName)];
    }
  });

  return list;
}

export default function useSelection<RecordType>(
  rowSelection: TableRowSelection<RecordType> | undefined,
  config: UseSelectionConfig<RecordType>,
): [TransformColumns<RecordType>, Set<Key>] {
  const {
    selections = true,
    preserveSelectedRowKeys,
    selectedRowKeys,
    getCheckboxProps,
    onChange,
    onSelectAll,
    onSelectInvert,
    onSelect,
    columnWidth: selectionColWidth,
    type: selectionType = 'checkbox',
    renderCell,
    checkStrictly = true,
    hideSelectAll,
    fixed,
  } = rowSelection || {};

  /**
   * 当没有设置 rowSelection 时，应该跳过 rowSelection 中耗时的操作
   */
  const shouldSkip = !rowSelection;

  const {
    prefix,
    tableLocale,
    data,
    pageData,
    getRecordByKey,
    getRowKey,
    getRowChildren,
    expandType,
    childrenColumnName,
  } = config;

  const [mergedSelectedKeys, setMergedSelectedKeys] = useControlledState(selectedRowKeys || [], selectedRowKeys);

  const lastSelectedKey = React.useRef<Key | null>(null);

  const recordsCacheRef = React.useRef(new Map<Key, RecordType>());

  const updateRecordsCache = useCallback(
    (keys: Key[]) => {
      if (preserveSelectedRowKeys) {
        const newCache = new Map<Key, RecordType>();
        keys.forEach(key => {
          let record = getRecordByKey(key);
          if (!record && recordsCacheRef.current.has(key)) {
            record = recordsCacheRef.current.get(key)!;
          }
          newCache.set(key, record);
        });
        recordsCacheRef.current = newCache;
      } else {
        recordsCacheRef.current = new Map<Key, RecordType>(
          keys.filter(getRecordByKey).map(k => [k, getRecordByKey(k)]),
        );
      }
    },
    [getRecordByKey, preserveSelectedRowKeys],
  );

  // update view when dataSource changed
  useEffect(() => {
    if (shouldSkip) {
      return;
    }
    updateRecordsCache(mergedSelectedKeys);
    const newMergedSelectedKeys = [...recordsCacheRef.current.keys()];
    if (difference(mergedSelectedKeys, newMergedSelectedKeys).length) {
      setMergedSelectedKeys(newMergedSelectedKeys);
    }
  }, [mergedSelectedKeys, setMergedSelectedKeys, updateRecordsCache, shouldSkip]);

  // Flatten Page Data
  const flattenedData = useMemo(
    () => (shouldSkip ? [] : flattenData(pageData, childrenColumnName)),
    [pageData, childrenColumnName, shouldSkip],
  );

  // Checkbox Props Map
  const checkboxPropsMap = useMemo(() => {
    const map = new Map<Key, Partial<CheckboxProps>>();
    flattenedData.forEach(record => {
      const key = getRowKey(record);
      const checkboxProps = (getCheckboxProps ? getCheckboxProps(record) : null) || {};
      map.set(key, checkboxProps);
    });
    return map;
  }, [flattenedData, getRowKey, getCheckboxProps]);

  const { keyEntities } = useMemo(
    () => (shouldSkip || checkStrictly ? { keyEntities: null } : convertDataToEntities(data, getRowKey)),
    [checkStrictly, data, getRowKey, shouldSkip],
  );
  // 不知道是干啥的。。。
  // const isCheckboxDisabled: GetCheckDisabled<RecordType> = useCallback(
  //   (r: RecordType) => Boolean(checkboxPropsMap.get(getRowKey(r))?.disabled),
  //   [checkboxPropsMap, getRowKey]
  // );

  const checkedCache = useMemo(() => {
    if (checkStrictly || !keyEntities) {
      return new Map();
    }
    const cache = checkedSet2Cache(keyEntities, array2Set(mergedSelectedKeys as string[]));
    return cache;
  }, [checkStrictly, keyEntities, mergedSelectedKeys]);

  const [derivedSelectedKeys, derivedHalfSelectedKeys] = useMemo(() => {
    if (checkStrictly || !keyEntities) {
      return [mergedSelectedKeys || [], []];
    }
    const keys: Key[] = [];
    const halfKeys: Key[] = [];
    for (const [key, value] of checkedCache.entries()) {
      if (value === CheckedType.Checked) {
        keys.push(key);
      }
      if (value === CheckedType.Indeterminate) {
        halfKeys.push(key);
      }
    }
    return [keys, halfKeys];
  }, [checkedCache, mergedSelectedKeys, checkStrictly, keyEntities]);

  const derivedSelectedKeySet: Set<Key> = useMemo(() => {
    const keys = selectionType === SELECTION_TYPE_RADIO ? derivedSelectedKeys.slice(0, 1) : derivedSelectedKeys;
    return new Set(keys);
  }, [derivedSelectedKeys, selectionType]);

  const derivedHalfSelectedKeySet: Set<Key> = useMemo(
    () => (selectionType === SELECTION_TYPE_RADIO ? new Set() : new Set(derivedHalfSelectedKeys)),
    [derivedHalfSelectedKeys, selectionType],
  );

  const setSelectedKeys = useCallback(
    (keys: Key[]) => {
      let availableKeys: Key[], records: RecordType[];
      updateRecordsCache(keys);
      if (preserveSelectedRowKeys) {
        // 用户设置了 preserveSelectedRowKeys 时，保留 key
        availableKeys = keys;
        records = keys.map(key => recordsCacheRef.current.get(key)!);
      } else {
        availableKeys = [];
        records = [];

        keys.forEach(key => {
          const record = getRecordByKey(key);
          if (record !== undefined) {
            availableKeys.push(key);
            records.push(record);
          }
        });
      }

      setMergedSelectedKeys(availableKeys, () => {
        isFunction(onChange) && onChange(availableKeys, records);
      });
    },
    [updateRecordsCache, preserveSelectedRowKeys, setMergedSelectedKeys, getRecordByKey, onChange],
  );

  // 用户手动勾选触发事件
  const triggerSelect = useCallback(
    (key: Key, selected: boolean, keys: Key[], event: Event) => {
      if (onSelect) {
        const rows = keys.map(k => getRecordByKey(k));
        onSelect(getRecordByKey(key), selected, rows, event);
      }
      setSelectedKeys(keys);
    },
    [onSelect, getRecordByKey, setSelectedKeys],
  );

  const keySet = useMemo(() => new Set(derivedSelectedKeySet), [derivedSelectedKeySet]);

  // 下拉菜单的配置项列表
  const mergedSelections = useMemo<SelectionItem[] | null>(() => {
    if (!selections || hideSelectAll) {
      return null;
    }
    const selectionList: INTERNAL_SELECTION_ITEM[] =
      selections === true ? [SELECTION_ALL, SELECTION_INVERT] : selections;
    if (selectionList.length === 0) {
      return null;
    }
    return selectionList.map((selection: INTERNAL_SELECTION_ITEM) => {
      if (selection === SELECTION_ALL) {
        return {
          key: 'all',
          text: tableLocale.selectionAll,
          onSelect() {
            setSelectedKeys(data.map((record, index) => getRowKey(record)));
          },
        };
      }
      if (selection === SELECTION_INVERT) {
        return {
          key: 'invert',
          text: tableLocale.selectInvert,
          onSelect() {
            const invertSet = (_keySet: Set<React.Key>, _pageData: readonly RecordType[]) => {
              const keys: React.Key[] = [];
              const childrenKeys: React.Key[] = [];
              _pageData.forEach(_data => {
                const key = getRowKey(_data);
                const children = getRowChildren?.(_data);

                if (checkStrictly) {
                  !_keySet.has(key) && keys.push(key);
                  if (children) {
                    const [_keys] = invertSet(_keySet, children);
                    keys.push(..._keys);
                  }
                } else {
                  if (children) {
                    const [_keys, _childrenKeys] = invertSet(_keySet, children);
                    _childrenKeys.length === children.length && keys.push(key);
                    keys.push(..._keys);
                    childrenKeys.push(..._childrenKeys);
                  } else {
                    !_keySet.has(key) && keys.push(key);
                  }
                }
              });
              return [keys, childrenKeys];
            };
            const [keys] = invertSet(keySet, pageData);
            if (onSelectInvert) {
              onSelectInvert(keys);
            }
            /**
             * 反选本页，仅替换本页数据
             * flattenedData: 打平以后的本页数据
             * flattenedPageKeys: 打平以后的本页数据的所有 key
             * keySet: 更新前已被选择的所有 key
             * keys: 反选操作计算后本页应该被选择的 key
             * finalSelectedKeys：用更新前已被选择的所有 key 去处掉本页的key，再将本页更新后新选择的 key 加进去
             *
             * eg: 当前共两页数据：[[1, 2, 3], [4, 5, 6]]，已经选择[2, 3, 4]，当前在第二页，点击反选
             *  flattenedPageKeys: [4, 5, 6]
             *  keySet: [2, 3, 4]
             *  keys: [5, 6]
             *  finalSelectedKeys：[2, 3, 4] - [4, 5, 6] + [5, 6] = [2, 3, 5, 6]
             */
            const flattenedPageKeys = flattenedData.map(d => getRowKey(d));
            const finalSelectedKeys = [...without([...keySet], ...flattenedPageKeys), ...keys];
            setSelectedKeys(finalSelectedKeys);
          },
        };
      }
      return selection as SelectionItem;
    });
  }, [
    checkStrictly,
    selections,
    hideSelectAll,
    tableLocale.selectionAll,
    tableLocale.selectInvert,
    setSelectedKeys,
    data,
    getRowKey,
    getRowChildren,
    pageData,
    onSelectInvert,
    keySet,
    flattenedData,
  ]);

  const transformColumns = useCallback(
    (columns: ColumnsType<RecordType>): ColumnsType<RecordType> => {
      if (!rowSelection) {
        return columns;
      }

      const recordKeys = flattenedData.map(getRowKey).filter(key => !checkboxPropsMap.get(key)!.disabled);

      const renderSelectionCell = (_: boolean, record: RecordType, index: number) => {
        const key = getRowKey(record);
        const checked = keySet.has(key);
        const checkboxProps = checkboxPropsMap.get(key);
        const indeterminate = derivedHalfSelectedKeySet.has(key);
        const mergedIndeterminate =
          expandType === 'nest' ? indeterminate : checkboxProps?.indeterminate ?? indeterminate;
        let onCellSelect;

        if (selectionType === SELECTION_TYPE_RADIO) {
          onCellSelect = ({ nativeEvent }: ChangeEvent<HTMLInputElement>) => {
            if (!keySet.has(key)) {
              triggerSelect(key, true, [key], nativeEvent);
            }
          };
        } else {
          onCellSelect = (e: ChangeEvent<HTMLInputElement>) => {
            // 鼠标点击、键盘敲击 checkBox，收到 pointerEvent 事件
            // 是否会收到其他事件？
            const nativeEvent = e.nativeEvent as PointerEvent;
            const shift = nativeEvent.shiftKey;

            if (checkStrictly) {
              if (shift) {
                // shift 连选
                const curSelectedKeys = selectWithShift(lastSelectedKey.current, key, recordKeys, keySet);
                triggerSelect(key, !checked, Array.from(curSelectedKeys), nativeEvent);
              } else {
                // 普通checkbox单选
                const curSelectedKeySet = new Set<Key>(keySet);
                if (checked) {
                  curSelectedKeySet.delete(key);
                } else {
                  curSelectedKeySet.add(key);
                }
                triggerSelect(key, !checked, Array.from(curSelectedKeySet), nativeEvent);
              }
            } else {
              // !checkStrictly
              if (!keyEntities) {
                return;
              }
              const entity = keyEntities[key];
              const checkedType = checkedCache.get(key);
              const checkedCacheClone = cloneMap(checkedCache);
              if (checkedType === CheckedType.Checked) {
                unCheck(entity, checkedCacheClone, checkStrictly ?? false, getRowKey);
              } else {
                check(entity, checkedCacheClone, checkStrictly ?? false, getRowKey);
              }
              const nextCheckedKeys = Object.entries(cache2CheckedSet(keyEntities, checkedCacheClone))
                .filter(([, v]) => v)
                .map(([k]) => k);
              triggerSelect(key, !checked, nextCheckedKeys, nativeEvent);
            }
            lastSelectedKey.current = key;
          };
        }

        const cellNode = (
          <SelectionCell
            {...checkboxProps}
            type={selectionType}
            checked={checked}
            indeterminate={mergedIndeterminate}
            onSelect={onCellSelect}
          />
        );
        if (renderCell) {
          return renderCell(checked, record, index, cellNode);
        }
        return cellNode;
      };

      const checkedCurrentAll = recordKeys.every((key: Key) => keySet.has(key));
      const checkedCurrentSome = recordKeys.some((key: Key) => keySet.has(key));
      const allDisabled = flattenedData.every((record: RecordType, index: number) => {
        const key = getRowKey(record);
        const checkboxProps = checkboxPropsMap.get(key) || {};
        return checkboxProps.disabled;
      });

      let defaultTitle: React.ReactNode;
      if (selectionType === SELECTION_TYPE_CHECKBOX && !hideSelectAll) {
        defaultTitle = (
          <SelectionTitle
            prefixCls={prefix}
            checked={!allDisabled && Boolean(flattenedData.length) && checkedCurrentAll}
            disabled={flattenedData.length === 0 || allDisabled}
            indeterminate={!checkedCurrentAll && checkedCurrentSome}
            recordKeys={recordKeys}
            mergedSelections={mergedSelections}
            onSelectAllChange={ev => {
              const changeKeys: Key[] = [];

              if (checkedCurrentAll) {
                recordKeys.forEach((key: Key) => {
                  keySet.delete(key);
                  changeKeys.push(key);
                });
              } else {
                recordKeys.forEach((key: Key) => {
                  if (!keySet.has(key)) {
                    keySet.add(key);
                    changeKeys.push(key);
                  }
                });
              }

              const keys: Key[] = Array.from(keySet);
              setSelectedKeys(keys);

              if (onSelectAll) {
                const rows = keys.map(k => getRecordByKey(k));
                const changedRows = changeKeys.map(k => getRecordByKey(k));
                onSelectAll(ev.target.checked, rows, changedRows);
              }
            }}
          />
        );
      }

      const selectionColumn = {
        width: selectionColWidth,
        className: `${prefix}-selection-column`,
        title: rowSelection.columnTitle || defaultTitle,
        render: renderSelectionCell,
        [INTERNAL_COL_DEFINE]: {
          className: `${prefix}-selection-col`,
        },
      };

      return [{ ...selectionColumn, fixed: fixed || columns[0]?.fixed }, ...columns];
    },
    [
      onSelectAll,
      rowSelection,
      flattenedData,
      getRowKey,
      selectionType,
      hideSelectAll,
      selectionColWidth,
      prefix,
      expandType,
      fixed,
      checkboxPropsMap,
      derivedHalfSelectedKeySet,
      renderCell,
      triggerSelect,
      checkStrictly,
      keyEntities,
      checkedCache,
      setSelectedKeys,
      getRecordByKey,
      mergedSelections,
      lastSelectedKey,
      keySet,
    ],
  );

  return [transformColumns, derivedSelectedKeySet];
}
