// @universe-design/react/src/table/hooks/useSelection/SelectionTitle.tsx
import React, { useState } from 'react';
import { Dropdown } from '../../../dropdown';
import { ExpandButton } from '../../../shared/components/expand-button';
import { Menu } from '../../../menu';
import type { CheckBoxChangeEvent } from '../../../checkbox';
import { Checkbox } from '../../../checkbox';
import type { Key, SelectionItem, GetPopupContainerInTable } from '../../interface';

interface SelectionConfig {
  prefixCls: string;
  checked: boolean;
  disabled: boolean;
  indeterminate: boolean;
  getPopupContainer?: GetPopupContainerInTable;
  recordKeys: Key[];
  mergedSelections: SelectionItem[] | null;
  onSelectAllChange: (e: CheckBoxChangeEvent) => void;
}

function SelectionTitle(props: SelectionConfig) {
  const {
    prefixCls,
    checked,
    disabled,
    indeterminate,
    getPopupContainer,
    recordKeys,
    onSelectAllChange,
    mergedSelections,
  } = props;

  let customizeSelections: React.ReactNode;

  if (mergedSelections) {
    const menu = (
      <Menu getPopupContainer={getPopupContainer}>
        {mergedSelections.map((selection, index) => {
          const { key, text, onSelect } = selection;
          return (
            <Menu.Item
              key={key || index}
              onClick={() => {
                if (onSelect) {
                  onSelect(recordKeys);
                }
              }}
            >
              {text}
            </Menu.Item>
          );
        })}
      </Menu>
    );

    // eslint-disable-next-line react-hooks/rules-of-hooks
    const [visible, setVisible] = useState(false);

    customizeSelections = (
      <Dropdown
        visible={visible}
        onVisibleChange={e => setVisible(e)}
        overlay={menu}
        getPopupContainer={getPopupContainer}
      >
        <ExpandButton
          isExpanded={visible}
          className={`${prefixCls}-selection-extra`}
          size="xxs"
          iconType="DownBoldOutlined"
        />
      </Dropdown>
    );
  }

  return (
    <div className={`${prefixCls}-selection`}>
      <Checkbox checked={checked} indeterminate={indeterminate} onChange={onSelectAllChange} disabled={disabled} />
      {customizeSelections}
    </div>
  );
}

export default SelectionTitle;
