// @universe-design/react/es/date-picker/cell/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React from 'react';
import clsx from 'clsx';
import { useCssScope } from '../../config-provider/hooks';
import { Range } from './Range.class';
import './index.less';

/**
 * DateCell 包含三个部分
 * - DateCellRangeBG，range 背景
 * - DateCellRangePreviewBG，range 预选背景
 * - DateCellInner，真正的日期元胞
 *
 * range 背景会有 start 和 end 的概念，会出现半圆形
 * range preview 和 range 是有叠加需求的
 */

const usePickerCellCssScope = () => useCssScope('picker__cell');
const DateCell = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    children,
    isSelected = false,
    isDisabled = false,
    isToday = false,
    isOutOfView = false,
    isRangePreview = false,
    isRange = false,
    className,
    prevCellStatus,
    nextCellStatus,
    mode = 'cell',
    colIndex,
    colNum,
    ...restProps
  } = _ref;
  const cssScope = usePickerCellCssScope();
  const rangeCssScope = useCssScope('picker__cell__range');
  if (mode === 'text') {
    return /*#__PURE__*/React.createElement("div", _extends({
      ref: ref,
      className: clsx(cssScope(), className)
    }, restProps), /*#__PURE__*/React.createElement("div", {
      className: clsx(cssScope('__inner'), cssScope('__inner-text'))
    }, children));
  }
  if (!prevCellStatus || !nextCellStatus || colIndex === undefined || colNum === undefined) {
    return null;
  }

  /**
   * 这段逻辑的解释在如下文档中
   * https://bytedance.feishu.cn/docx/doxcn9kqBZVTrbolR9ilVUzsR1X
   */

  // 全局截断

  const isRowEnd = colIndex === colNum - 1;
  const {
    cellColor: rangeCellColor,
    beforeColor: rangeBeforeColor,
    afterColor: rangeAfterColor
  } = new Range(prevCellStatus, {
    isSelected,
    isDisabled,
    isToday,
    isOutOfView,
    isRangePreview,
    isRange
  }, nextCellStatus);
  const rangeBeforeBGCls = (() => {
    if (colIndex === 0) {
      return;
    }
    return rangeBeforeColor ? rangeCssScope(`-bg-${rangeBeforeColor}`) : undefined;
  })();
  const rangeAfterBGCls = (() => {
    if (isRowEnd) {
      return;
    }
    return rangeAfterColor ? rangeCssScope(`-bg-${rangeAfterColor}`) : undefined;
  })();

  // Range 跟前面一段相接的部分
  const beforeRangeNode = /*#__PURE__*/React.createElement("div", {
    className: clsx(rangeCssScope('__before'), rangeBeforeBGCls)
  });

  // Range 跟后面一段相接的部分
  const afterRangeNode = /*#__PURE__*/React.createElement("div", {
    className: clsx(rangeCssScope('__after'), rangeAfterBGCls)
  });
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("div", _extends({
    ref: ref,
    className: clsx(cssScope(), isSelected && cssScope('-selected'), isOutOfView && cssScope('-outOfView'), isDisabled && cssScope('-disabled'), isToday && cssScope('-today'), isRange && cssScope('-range'), isRangePreview && cssScope('-rangePreview'), className)
  }, restProps), beforeRangeNode, afterRangeNode, /*#__PURE__*/React.createElement("div", {
    className: clsx(cssScope('__range'), rangeCellColor ? rangeCssScope(`-bg-${rangeCellColor}`) : undefined)
  }, /*#__PURE__*/React.createElement("div", {
    className: cssScope('__inner')
  }, children))), !isRowEnd && /*#__PURE__*/React.createElement("div", {
    className: clsx(cssScope('__space'), rangeAfterBGCls)
  }));
});
export const useDateCellInteractiveAreaCls = () => usePickerCellCssScope()('-interactive-area');
export default DateCell;
