// @universe-design/react/es/date-picker/cell/Range.class.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
var RangeType = /*#__PURE__*/function (RangeType) {
  RangeType[RangeType["OutOfView"] = 0] = "OutOfView";
  RangeType[RangeType["InView"] = 1] = "InView";
  RangeType[RangeType["Selected"] = 2] = "Selected";
  RangeType[RangeType["OutOfSelected"] = 3] = "OutOfSelected";
  RangeType[RangeType["Preview"] = 4] = "Preview";
  RangeType[RangeType["Disabled"] = 5] = "Disabled";
  return RangeType;
}(RangeType || {});
export class Range {
  constructor(_prev, _cur, _next) {
    _defineProperty(this, "_prev", void 0);
    _defineProperty(this, "_cur", void 0);
    _defineProperty(this, "_next", void 0);
    _defineProperty(this, "_startEnd", void 0);
    this._prev = _prev;
    this._cur = _cur;
    this._next = _next;
    this._startEnd = new Map([[RangeType.OutOfView, {
      start: Boolean(this._cur.isOutOfView && !this._prev.isOutOfView),
      end: Boolean(this._cur.isOutOfView && !this._next.isOutOfView),
      in: Boolean(this._cur.isOutOfView)
    }], [RangeType.InView, {
      start: Boolean(!this._cur.isOutOfView && this._prev.isOutOfView),
      end: Boolean(!this._cur.isOutOfView && this._next.isOutOfView),
      in: !this._cur.isOutOfView
    }], [RangeType.Selected, {
      start: Boolean(this._cur.isRange && !this._prev.isRange),
      end: Boolean(this._cur.isRange && !this._next.isRange),
      in: Boolean(this._cur.isRange)
    }], [RangeType.OutOfSelected, {
      start: Boolean(!this._cur.isRange && this._prev.isRange),
      end: Boolean(!this._cur.isRange && this._next.isRange),
      in: !this._cur.isRange
    }], [RangeType.Preview, {
      start: Boolean(this._cur.isRangePreview && !this._prev.isRangePreview),
      end: Boolean(this._cur.isRangePreview && !this._next.isRangePreview),
      in: Boolean(this._cur.isRangePreview)
    }], [RangeType.Disabled, {
      start: Boolean(this._cur.isDisabled && !this._prev.isDisabled),
      end: Boolean(this._cur.isDisabled && !this._next.isDisabled),
      in: Boolean(this._cur.isDisabled)
    }]]);
  }
  get cellColor() {
    var _this$_startEnd$get;
    const typeSet = new Set();
    for (const [rangeType, se] of this._startEnd.entries()) {
      if (se.in) {
        typeSet.add(rangeType);
      }
    }
    if ((_this$_startEnd$get = this._startEnd.get(RangeType.OutOfView)) !== null && _this$_startEnd$get !== void 0 && _this$_startEnd$get.in) {
      return;
    }
    return Range.rangeTypes2rangeColor(typeSet);
  }
  get beforeColor() {
    var _this$_startEnd$get2, _this$_startEnd$get3, _this$_startEnd$get4, _this$_startEnd$get5;
    const typeSet = new Set();
    for (const [rangeType, se] of this._startEnd.entries()) {
      if (se.in && !se.start) {
        typeSet.add(rangeType);
      }
    }
    if ((_this$_startEnd$get2 = this._startEnd.get(RangeType.OutOfView)) !== null && _this$_startEnd$get2 !== void 0 && _this$_startEnd$get2.in || (_this$_startEnd$get3 = this._startEnd.get(RangeType.InView)) !== null && _this$_startEnd$get3 !== void 0 && _this$_startEnd$get3.start) {
      return;
    }
    if ((_this$_startEnd$get4 = this._startEnd.get(RangeType.Selected)) !== null && _this$_startEnd$get4 !== void 0 && _this$_startEnd$get4.start || (_this$_startEnd$get5 = this._startEnd.get(RangeType.OutOfSelected)) !== null && _this$_startEnd$get5 !== void 0 && _this$_startEnd$get5.start) {
      typeSet.delete(RangeType.Disabled);
    }
    return Range.rangeTypes2rangeColor(typeSet);
  }
  get afterColor() {
    var _this$_startEnd$get6, _this$_startEnd$get7, _this$_startEnd$get8, _this$_startEnd$get9;
    const typeSet = new Set();
    for (const [rangeType, se] of this._startEnd.entries()) {
      if (se.in && !se.end) {
        typeSet.add(rangeType);
      }
    }
    // 特化逻辑，view range 开头结尾前后需要截断所有
    if ((_this$_startEnd$get6 = this._startEnd.get(RangeType.OutOfView)) !== null && _this$_startEnd$get6 !== void 0 && _this$_startEnd$get6.in || (_this$_startEnd$get7 = this._startEnd.get(RangeType.InView)) !== null && _this$_startEnd$get7 !== void 0 && _this$_startEnd$get7.end) {
      return;
    }
    // 特化逻辑，selected range 开头结尾的前后需要截断 disabled
    if ((_this$_startEnd$get8 = this._startEnd.get(RangeType.Selected)) !== null && _this$_startEnd$get8 !== void 0 && _this$_startEnd$get8.end || (_this$_startEnd$get9 = this._startEnd.get(RangeType.OutOfSelected)) !== null && _this$_startEnd$get9 !== void 0 && _this$_startEnd$get9.end) {
      typeSet.delete(RangeType.Disabled);
    }
    return Range.rangeTypes2rangeColor(typeSet);
  }
  static rangeTypes2rangeColor(typeSet) {
    if (typeSet.has(RangeType.Disabled)) {
      return 'disabled';
    }
    if (typeSet.has(RangeType.Preview) && typeSet.has(RangeType.Selected)) {
      return 'selected-preview';
    }
    if (typeSet.has(RangeType.Preview)) {
      return 'preview';
    }
    if (typeSet.has(RangeType.Selected)) {
      return 'selected';
    }
  }
}
