// @universe-design/react-ai/lib-esm/star-thinking-animation/index.js
//.u18ujvr1{top:85.2%;}
import u874iyf from "./../.stylex/u874iyf";
//.ub5ljux{animation:ai__star-thinking__sub-lb__animation 2.15s cubic-bezier(0.45, 0, 0.55, 1) infinite;}
import uzqxen4 from "./../.stylex/uzqxen4";
//.uwodgp6{left:84.4%;}
import uxirdmd from "./../.stylex/uxirdmd";
//.u1grlrm6{animation:ai__star-thinking__sub-rt__animation 2.15s cubic-bezier(0.45, 0, 0.55, 1) infinite;}
import uupcts6 from "./../.stylex/uupcts6";
//.ukpg2sn{left:15.6%;}
import uivn5xe from "./../.stylex/uivn5xe";
//.u1awxekb{top:14.9%;}
import u18zhii8 from "./../.stylex/u18zhii8";
//.uczczkq{animation:ai__star-thinking__sub-lt__animation 2.15s cubic-bezier(0.45, 0, 0.55, 1) infinite;}
import u14enhrg from "./../.stylex/u14enhrg";
//.u1euv43f{position:absolute;}
import u1u1p7mj from "./../.stylex/u1u1p7mj";
//.u1ktbn1t{left:50%;}
import u1jmni06 from "./../.stylex/u1jmni06";
//.u1i1t8d1{top:50%;}
import ufibavv from "./../.stylex/ufibavv";
//.uou3m91{font-size:var(--svg-size);}
import uqut2en from "./../.stylex/uqut2en";
//.u17ctas6{animation:ai__star-thinking__main__animation 2.15s cubic-bezier(0.37, 0, 0.63, 1) infinite;}
import u1k0zs9m from "./../.stylex/u1k0zs9m";
//.uparq7u{--font-size:1em;}
import u73r7in from "./../.stylex/u73r7in";
//.u14t3ns0{display:inline-block;}
import u8wgnjj from "./../.stylex/u8wgnjj";
//.u10pi13n{position:relative;}
import u5jxv71 from "./../.stylex/u5jxv71";
//.u1cyak3s{width:var(--svg-size);}
import u1a7olnx from "./../.stylex/u1a7olnx";
//.u18fh3uv{height:var(--svg-size);}
import u4hyup3 from "./../.stylex/u4hyup3";
//.u15xaump{vertical-align:calc(0.36 * var(--font-size) - 0.5 * var(--svg-size));}
import u1ma4v2k from "./../.stylex/u1ma4v2k";
import React from "react";
/* eslint-disable react/no-unknown-property */
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { makeStylexStyles, jsx } from '@universe-design/stylex/react';
import MeetingAiOutlined from '@universe-design/icons-react/MeetingAiOutlined';
import MeetingAiFilled from '@universe-design/icons-react/MeetingAiFilled';
import MeetingAiTestFilled from '@universe-design/icons-react/MeetingAiTestFilled';
import stylex from '@universe-design/stylex';
import './index.less';
const useStarThinkingAnimationStyles = makeStylexStyles(_ref => {
  let {
    prefix
  } = _ref;
  return {
    root: {
      "props": {
        "B3yy3ex": u73r7in,
        "mc9l5x": u8wgnjj,
        "qhf8xq": u5jxv71,
        "a9b677": u1a7olnx,
        "Bqenvij": u4hyup3,
        "ha4doy": u1ma4v2k
      },
      "alias": `${prefix}__star-thinking`
    },
    main: {
      "props": {
        "qhf8xq": u1u1p7mj,
        "oyh7mz": u1jmni06,
        "Bhzewxz": ufibavv,
        "Be2twd7": uqut2en,
        "Ba009jy": u1k0zs9m
      },
      "alias": `${prefix}__star-thinking__main`
    },
    sub: {
      base: {
        "props": {
          "qhf8xq": u1u1p7mj,
          "Be2twd7": uqut2en
        },
        "alias": `${prefix}__star-thinking__sub`
      },
      lt: {
        "props": {
          "oyh7mz": uivn5xe,
          "Bhzewxz": u18zhii8,
          "Ba009jy": u14enhrg
        },
        "alias": `${prefix}__star-thinking__sub-lt`
      },
      rt: {
        "props": {
          "oyh7mz": uxirdmd,
          "Bhzewxz": u18zhii8,
          "Ba009jy": uupcts6
        },
        "alias": `${prefix}__star-thinking__sub-rt`
      },
      lb: {
        "props": {
          "oyh7mz": uivn5xe,
          "Bhzewxz": u874iyf,
          "Ba009jy": uzqxen4
        },
        "alias": `${prefix}__star-thinking__sub-lb`
      }
    }
  };
});

/**
 * 星型思考动画图标组件入参
 *
 * @public
 */

/**
 * 星型思考动画图标
 *
 * @public
 */
export const StarThinkingAnimation = props => {
  const {
    type = 'outlined',
    size = 24,
    style,
    className,
    sx
  } = props;
  const styles = useStarThinkingAnimationStyles({
    prefix: 'ai'
  });
  return jsx("div", {
    sx: [styles.root, sx],
    style: {
      '--svg-size': typeof size === 'number' ? `${size}px` : size,
      ...style
    },
    className: className
  }, type === 'filled' ? jsx(MeetingAiFilled, {
    className: stylex(styles.main)
  }) : jsx(MeetingAiOutlined, {
    className: stylex(styles.main)
  }), jsx(MeetingAiTestFilled, {
    className: stylex(styles.sub.base, styles.sub.lt)
  }), jsx(MeetingAiTestFilled, {
    className: stylex(styles.sub.base, styles.sub.rt)
  }), jsx(MeetingAiTestFilled, {
    className: stylex(styles.sub.base, styles.sub.lb)
  }));
};
