// @universe-design/react/lib/select/context/useSelectInput.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useSelectInput = useSelectInput;
var React = _interopRequireWildcard(require("react"));
var _util = require("../../util");
var _reactUtilities = require("@universe-design/react-utilities");
var _internal = require("../internal.interface");
function getSeparatedContent(text, tokens) {
  if (!tokens || !tokens.length) {
    return null;
  }
  let match = false;
  function separate(str, _ref) {
    let [token, ...restTokens] = _ref;
    if (!token) {
      return [str];
    }
    const list = str.split(token);
    match = match || list.length > 1;
    return list.reduce((prevList, unitStr) => [...prevList, ...separate(unitStr, restTokens)], []).filter(unit => unit);
  }
  const list = separate(text, tokens);
  return match ? list : null;
}
function useSelectInput(_ref2) {
  let {
    multiple,
    tokenSeparators,
    inputValue,
    setInputValue,
    set,
    deselectLast,
    setOverlayVisible,
    listRef,
    onKeyDown,
    onKeyUp,
    overlayVisible,
    setInputting,
    getOption
  } = _ref2;
  // When paste come, ignore next onChange
  const pastedTextRef = React.useRef(null);
  const compositionStatusRef = React.useRef(false);
  const tokenWithEnter = React.useMemo(() => (tokenSeparators || []).some(tokenSeparator => ['\n', '\r\n'].includes(tokenSeparator)), [tokenSeparators]);

  // ============================ Keyboard ============================

  // KeyUp

  return {
    inputValue,
    onInputChange: e => {
      let {
        target: {
          value: searchText
        }
      } = e;

      // Pasted text should replace back to origin content
      if (tokenWithEnter && pastedTextRef.current && /[\r\n]/.test(pastedTextRef.current)) {
        // CRLF will be treated as a single space for input element
        const replacedText = pastedTextRef.current.replace(/\r\n/g, ' ').replace(/[\r\n]/g, ' ');
        searchText = searchText.replace(replacedText, pastedTextRef.current);
      }
      pastedTextRef.current = null;
      const patchLabels = compositionStatusRef.current ? null : getSeparatedContent(searchText, tokenSeparators);
      if (!multiple) {
        /* empty */
      } else if (patchLabels) {
        searchText = '';
        set(patchLabels.map(label => {
          const item = getOption(label);
          return item === null || item === void 0 ? void 0 : item.value;
        }).filter(_util.isNonNullable));
        return;
      }
      setInputValue(searchText, {
        action: _internal.SelectSearchAction.Input
      });
    },
    onInputCompositionStart: () => {
      compositionStatusRef.current = true;
    },
    onInputCompositionEnd: () => {
      compositionStatusRef.current = false;
    },
    onInputPaste: e => {
      const {
        clipboardData
      } = e;
      const value = clipboardData.getData('text');
      pastedTextRef.current = value;
    },
    onInternalKeyDown: e => {
      (0, _util.isFunction)(onKeyDown) && onKeyDown(e);
      if (overlayVisible && listRef.current) {
        listRef.current.onKeyDown(e);
      }

      // 允许 overlay component 阻止默认行为
      if (e.defaultPrevented) {
        return;
      }
      setInputting(true);
      if (multiple && e.key === _reactUtilities.KeyboardKey.Backspace) {
        // DeleteTag
        if (inputValue === '') {
          deselectLast();
        }
      }
      if ([
      // 禁用浏览器 input 的光标默认行为
      _reactUtilities.KeyboardKey.Up, _reactUtilities.KeyboardKey.Down, _reactUtilities.KeyboardKey.ArrowUp, _reactUtilities.KeyboardKey.ArrowDown,
      // 禁用浏览器 enter 触发 click 事件，当 trigger element 为 button 时会有这种情况出现
      _reactUtilities.KeyboardKey.Enter].includes(e.key)) {
        e.preventDefault();
      }

      // Enter 事件处理不在这里，因为 Enter 的关闭逻辑比较难统一
      // ESC 关闭
      if (e.key === _reactUtilities.KeyboardKey.Escape) {
        // overlay 操作关闭 overlay
        setOverlayVisible(false);
        return;
      }

      // 禁用一些不需要在 input 这边操作的键，其他键都可以触发 overlay 打开
      if (![_reactUtilities.KeyboardKey.Shift,
      // 组合键
      _reactUtilities.KeyboardKey.Tab,
      // 浏览器默认行为切换 input 选中态
      _reactUtilities.KeyboardKey.Escape].includes(e.key)) {
        setOverlayVisible(true);
      }
    },
    onInternalKeyUp: e => {
      (0, _util.isFunction)(onKeyUp) && onKeyUp(e);
      if (overlayVisible && listRef.current) {
        listRef.current.onKeyUp(e);
      }
    },
    compositionStatusRef
  };
}
