// @universe-design/react/lib/select/context/useSelectFocus.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useSelectFocus = useSelectFocus;
var _react = require("react");
var _reactUtilities = require("@universe-design/react-utilities");
var _hooks = require("../../hooks");
var _util = require("../../util");
var _internal = require("../internal.interface");
function callDelay(cb) {
  let duration = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 250;
  let timeoutID = null;
  const cancel = () => {
    if (timeoutID !== null) {
      window.clearTimeout(timeoutID);
      timeoutID = null;
    }
  };
  return {
    call: function () {
      for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }
      cancel();
      timeoutID = window.setTimeout(() => cb(...args), duration);
    },
    cancel
  };
}

/**
 * use select focus management
 * 使用 input 的 focus 状态来代替整个组件的 focus 状态
 */
function useSelectFocus(_ref) {
  let {
    disabled,
    autoFocus,
    onFocus,
    onBlur,
    setOverlayVisible,
    overlayVisible,
    setInputValue,
    setInputting
  } = _ref;
  const _setOverlayVisible = (0, _hooks.useEventCallback)(setOverlayVisible);
  const _onBlur = (0, _hooks.useEventCallback)(onBlur);
  const inputRef = (0, _react.useRef)(null);
  const focusRef = (0, _react.useRef)(false);
  const [focusState, setFocusState] = (0, _react.useState)(false);
  const {
    call: callCloseOverlayDelay,
    cancel: cancelDelayCloseOverlay
  } = (0, _reactUtilities.useConstant)(() => callDelay(e => {
    if ((0, _util.isFunction)(_onBlur)) {
      _onBlur(e);
    }
    _setOverlayVisible(false);
    setInputValue('', {
      action: _internal.SelectSearchAction.Blur
    });
  }));

  /**
   * container focus 事件
   *
   * 接受一种不带 event onInputFocus，在 dropdown 点击的时候触发，避免 dropdown 关闭
   */
  const onContainerFocus = (0, _react.useCallback)(e => {
    cancelDelayCloseOverlay();
    if ((0, _util.isFunction)(onFocus) && !focusRef.current && e) {
      onFocus(e);
    }
    setFocusState(true);
    focusRef.current = true;
  }, [onFocus, cancelDelayCloseOverlay]);
  const onContainerBlur = (0, _react.useCallback)(e => {
    // 这段放在最前面是为了直接从 dropdown input blur 时 select 可以得到正确的 focus 样式
    setFocusState(false);
    focusRef.current = false;
    if (e.target !== inputRef.current) {
      return;
    }
    // 我们有一个逻辑是点击任何不是 input 的都要重新聚焦 input，这个时候 input 会有短暂的 blur 状态，所以我们需要区分正常 blur 和这种短时间的 blur
    callCloseOverlayDelay(e);
  }, [callCloseOverlayDelay]);
  (0, _hooks.useEffectOnce)(() => {
    // init focus，如果初始 overlay visible 或者 auto focus，则组件为 focus 状态
    if (autoFocus || overlayVisible) {
      var _inputRef$current;
      setFocusState(true);
      setInputting(true);
      (_inputRef$current = inputRef.current) === null || _inputRef$current === void 0 ? void 0 : _inputRef$current.focus();
    }
  });
  const imperativeFocus = (0, _react.useCallback)(options => {
    var _inputRef$current2;
    if (disabled) {
      return;
    }
    // 用户手动触发 focus 时，设置为输入中
    setInputting(true);
    (_inputRef$current2 = inputRef.current) === null || _inputRef$current2 === void 0 ? void 0 : _inputRef$current2.focus(options);
  }, [disabled, setInputting]);
  const blur = (0, _react.useCallback)(() => {
    var _inputRef$current3;
    if (disabled) {
      return;
    }
    (_inputRef$current3 = inputRef.current) === null || _inputRef$current3 === void 0 ? void 0 : _inputRef$current3.blur();
  }, [disabled]);
  return {
    focus: imperativeFocus,
    blur,
    focusState,
    inputRef,
    focusRef,
    onContainerFocus,
    onContainerBlur,
    cancelDelayCloseOverlay
  };
}
