// @universe-design/react/lib/select/context/useSelectAll.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = useSelectAll;
var _react = _interopRequireWildcard(require("react"));
var _internal = require("../internal.interface");
/**
 *  全选的相关逻辑：
 *  1. 如果没有 options: unChecked, disabled
 *  2. disabled & unChecked 的option, 不参与数量统计
 *  3. 如果 options 全为 disabled 态， 全选操作禁用，如果里面有 checked 态 option, 全选选中
 *
 *  最终状态：
 *  disabled:
 *    选项全部为 disabled 或 选项个数为 0
 *  checked:
 *    选中的个数不为0， 并且所有可选的选项已选
 *  unChecked
 *    其他情况
 *  总数：
 *    可选选项及disabled 但选中的个数
 */
function useSelectAll(args) {
  const {
    flattenOptions,
    valueSet,
    set
  } = args;
  const options = flattenOptions.filter(item => !item.group).map(op => op.option);
  const {
    isSelectAllDisabled,
    selectAllCount,
    hasSelectedAll
  } = (0, _react.useMemo)(() => {
    if (!options.length) {
      return {
        isSelectAllDisabled: true,
        selectAllCount: 0,
        hasSelectedAll: false
      };
    }
    // 底部显示的计数
    let count = 0;
    // 已经勾选的个数，包含禁用和非禁用态
    let selectedCount = 0;
    // 是否有可选但未选中的选项
    let hasUnCheckedOption = false;
    // 全选是否禁用
    let isDisabled = true;
    options.forEach(op => {
      if (!op.disabled) {
        isDisabled = false;
      }
      const isSelected = valueSet.has(op.value);
      if (isSelected) {
        selectedCount++;
      }
      if (isSelected || !op.disabled) {
        count++;
      }
      if (!isSelected && !op.disabled) {
        hasUnCheckedOption = true;
      }
    });
    return {
      isSelectAllDisabled: isDisabled,
      selectAllCount: count,
      hasSelectedAll: !hasUnCheckedOption && selectedCount > 0
    };
  }, [options, valueSet]);
  const onToggleSelectAll = _react.default.useCallback(() => {
    const _valueSet = new Set(valueSet);
    const toggleItems = options.filter(op => !op.disabled).map(op => op.value);
    for (const v of toggleItems) {
      if (!hasSelectedAll) {
        _valueSet.add(v);
      } else {
        _valueSet.delete(v);
      }
    }
    const action = {
      type: hasSelectedAll ? _internal.SelectActionType.DeSelectAll : _internal.SelectActionType.SelectAll,
      payload: toggleItems
    };
    set([..._valueSet], {
      action
    });
  }, [hasSelectedAll, options, valueSet, set]);
  return {
    hasSelectedAll,
    onToggleSelectAll,
    isSelectAllDisabled,
    selectAllCount
  };
}
