// @universe-design/react/lib/date-picker/Context.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useDatePickerContextSelector = exports.useDatePickerContext = exports.DatePickerContextProvider = void 0;
var _useContextSelector = require("use-context-selector");
var _util = require("../util");
const DatePickerContext = (0, _useContextSelector.createContext)({
  mode: 'date',
  setMode: () => process.env.NODE_ENV !== "production" ? (0, _util.warning)(false, (0, _util.devWarning)('DatePicker', 'You are using the default setMode from context')) : void 0,
  setValue: () => process.env.NODE_ENV !== "production" ? (0, _util.warning)(false, (0, _util.devWarning)('DatePicker', 'You are using the default setValue from context')) : void 0,
  setOpened: () => process.env.NODE_ENV !== "production" ? (0, _util.warning)(false, (0, _util.devWarning)('DatePicker', 'You are using the default setOpened from context')) : void 0,
  showToday: true,
  showTime: false,
  showNow: false,
  needConfirm: false
});
const useDatePickerContext = () => (0, _useContextSelector.useContext)(DatePickerContext);
exports.useDatePickerContext = useDatePickerContext;
const useDatePickerContextSelector = selector => (0, _useContextSelector.useContextSelector)(DatePickerContext, selector);
exports.useDatePickerContextSelector = useDatePickerContextSelector;
const {
  Provider: DatePickerContextProvider
} = DatePickerContext;
exports.DatePickerContextProvider = DatePickerContextProvider;
