// @universe-design/react/lib/date-picker/DatePicker.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DatePicker = DatePicker;
var _CalendarOutlined = _interopRequireDefault(require("@universe-design/icons-react/CalendarOutlined"));
var _clsx = _interopRequireDefault(require("clsx"));
var _noop = _interopRequireDefault(require("lodash/noop"));
var _moment = _interopRequireDefault(require("moment"));
var _react = _interopRequireWildcard(require("react"));
var _reactUtilities = require("@universe-design/react-utilities");
var _hooks = require("../config-provider/hooks");
var _useControlled = _interopRequireDefault(require("../hooks/useControlled"));
var _useSyncState = _interopRequireDefault(require("../hooks/useSyncState"));
var _Input = _interopRequireDefault(require("../shared/components/Input"));
var _List = require("../shared/components/list/List");
var _util = require("../util");
var _composeRefs = _interopRequireDefault(require("../util/composeRefs"));
var _Context = require("./Context");
var _picker = _interopRequireWildcard(require("./panels/picker"));
var _date = require("./utils/date");
var _ui = require("./utils/ui");
var _moment2 = _interopRequireDefault(require("../shared/components/picker/generate/moment"));
require("./style");
var _usePopperSidecar = require("../config-provider/hooks/usePopperSidecar");
var _SlideUpDownTransition = require("../shared/components/transition/SlideUpDownTransition");
var _useClickOutsideEffect = _interopRequireDefault(require("../shared/components/picker/hooks/useClickOutsideEffect"));
var _inputWrapper = _interopRequireDefault(require("../shared/components/picker/Comp/input-wrapper"));
var _useFormatAndParser = _interopRequireDefault(require("./hooks/useFormatAndParser"));
var _disabled = require("./utils/disabled");
var _reactProvider = require("@universe-design/react-provider");
// nolint-next-line: long_function
function DatePicker(props) {
  const cssScope = (0, _hooks.useCssScope)('picker');
  const prefixCls = cssScope();
  const locale = (0, _hooks.useLocaleReceiver)('datePicker');
  const defaultSize = (0, _hooks.useComponentSize)();
  const inputGroupSize = (0, _hooks.useInputGroupSize)();
  const {
    style,
    id,
    showTime = false,
    picker = 'date',
    mode = picker,
    disabledDate: disabledDateProp,
    minDate,
    maxDate,
    className,
    allowClear,
    autoFocus,
    disabled,
    // eslint-disable-next-line no-nested-ternary
    placeholder = picker === 'date' ? locale.placeholder : mode === 'month' ? locale.monthPlaceholder : locale.yearPlaceholder,
    inputReadOnly,
    format: formatProp,
    size,
    error,
    triggerElement,
    parse,
    onKeyDown = _noop.default,
    onMouseDown = _noop.default,
    onMouseUp = _noop.default,
    onClick = _noop.default,
    suffixIcon = /*#__PURE__*/_react.default.createElement(_CalendarOutlined.default, null),
    renderExtraFooter,
    showToday = !showTime && picker === 'date',
    getPopupContainer,
    dropdownClassName,
    popupStyle,
    startOfWeek = 0,
    dateRender,
    onClear,
    onSelect,
    defaultPickerValue,
    defaultOpen = false,
    value: valueProp,
    defaultValue,
    onChange,
    onBlur,
    needConfirm = Boolean(showTime),
    showNow = Boolean(showTime),
    hideWeekNumbers,
    janX
  } = props;
  const disabledDate = (0, _react.useMemo)(() => {
    if (disabledDateProp === undefined) {
      return undefined;
    }
    const cache = new Map();
    return function (date) {
      const cacheKey = date.format('YYYYMMDD');
      if (cache.has(cacheKey)) {
        return cache.get(cacheKey);
      }
      const isDisabled = disabledDateProp(date);
      cache.set(cacheKey, isDisabled);
      return isDisabled;
    };
  }, [disabledDateProp]);
  const direction = (0, _hooks.useDirection)();
  const Popper = (0, _usePopperSidecar.usePopperSidecar)();
  const cmpSize = (0, _util.getDefaultSizeAbbr)(size || inputGroupSize || defaultSize);
  const showTimeSelection = Boolean(showTime);
  const {
    formatValue,
    parseValue,
    format
  } = (0, _useFormatAndParser.default)({
    mode,
    format: formatProp,
    parse,
    showTime: showTimeSelection
  });
  process.env.NODE_ENV !== "production" ? (0, _util.warning)(!(showNow && showToday), (0, _util.devWarning)('datePicker', 'should not set showNow and showToday at same time')) : void 0;
  const dateToText = _react.default.useCallback(date => formatValue(date), [formatValue]);
  const [value, setValue] = (0, _reactUtilities.useControlledState)(
  // 如果 valueProp 是 moment 形式，采用 valueProp
  // 如果 valueProp 是 undefined 或其他，统一采用 undefined 形式
  _moment.default.isMoment(defaultValue) && defaultValue.isValid() ? defaultValue : undefined, _moment.default.isMoment(valueProp) && valueProp.isValid() ? valueProp : undefined, {
    onChange,
    isEqual: _picker.isDateValueEqual,
    isControlledOverride: 'value' in props
  });

  /**
   * 当前视图可见的有效date
   */
  const [viewValue, setViewValue] = (0, _useSyncState.default)(value, {
    isEqual: _picker.isDateValueEqual
  });
  const validText = dateToText(viewValue);

  /**
   * 输入框的值
   */
  const [text, setText] = (0, _useSyncState.default)(validText);

  /**
   * hover相关的值
   */
  const [hoverDate, setHoverDate] = (0, _react.useState)();
  const hoverText = dateToText(hoverDate);
  const [focused, setFocused] = (0, _react.useState)(false);
  const [open, setOpened] = (0, _useControlled.default)(props, {
    defaultValuePropName: '',
    valuePropName: 'open',
    onChangePropName: 'onOpenChange',
    defaultValue: defaultOpen
  });
  const panelDivRef = _react.default.useRef(null);
  const inputDivRef = _react.default.useRef(null);
  const inputRef = (0, _react.useRef)(null);
  const triggerElementRef = _react.default.useRef(null);
  const rootCls = (0, _clsx.default)(cssScope(), {
    [cssScope(`--size-${cmpSize}`)]: cmpSize,
    [cssScope('--error')]: error,
    [cssScope('--focus')]: focused,
    [cssScope('--disabled')]: disabled
  });
  const ctxGetPopupContainer = (0, _reactProvider.usePopupPortalContainer)();
  const confirm = (0, _react.useCallback)(newValue => {
    if (!(0, _date.isEqual)(value, newValue, mode, showTimeSelection)) {
      setValue(newValue);
    }

    // 暂时同步value值，不然受控模式下，可能出现问题
    setViewValue(value);
    setText(dateToText(value));
    setHoverDate(undefined);
    setOpened(false);
  }, [value, mode, showTimeSelection, setViewValue, setText, dateToText, setHoverDate, setOpened, setValue]);
  const cancel = (0, _react.useCallback)(() => {
    setOpened(false);
    setViewValue(value);
    setText(dateToText(value));
  }, [dateToText, setOpened, setText, setViewValue, value]);
  function handleKeyDown(e) {
    onKeyDown(e);
    if (e.key !== 'Tab' && !open) {
      setOpened(true);
      return;
    }

    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    e.keyCode === 13 && confirm(viewValue);
  }
  const handleInputChange = e => {
    setHoverDate(undefined);
    const editText = e.target.value;
    setText(editText);
    const date = parseValue(editText);
    if (date !== null && date !== void 0 && date.isValid() && !(0, _disabled.isInputDateDisabled)(date, {
      minDate,
      maxDate,
      disabledDate
    })) {
      setViewValue(date);
    }
  };
  const handleClear = () => {
    setValue(undefined);
    // 暂时同步value值，不然受控模式下，可能出现问题
    setViewValue(value);
    setOpened(false);
    onClear === null || onClear === void 0 ? void 0 : onClear();
  };
  const handleMouseUp = e => {
    if (inputRef.current) {
      inputRef.current.focus();
    }
    setOpened(true);
    onMouseUp(e);
  };
  const handleFocus = e => {
    setFocused(true);
  };
  const handleBlur = e => {
    onBlur === null || onBlur === void 0 ? void 0 : onBlur(e);
    setFocused(false);
    needConfirm ? cancel() : confirm(viewValue);
  };

  /**
   * 点击弹窗外部行为：
   *    1. showTime
   *       1. confirmButton: 取消
   *       2. 非 confirmButton：提交
   *    2. !showTime
   *        1. confirmButton: 取消
   *        2. 非 confirmButton：取消
   */
  (0, _useClickOutsideEffect.default)((0, _react.useMemo)(() => [panelDivRef, inputDivRef, triggerElementRef], []), open, (0, _react.useCallback)(() => {
    !needConfirm && showTime ? confirm(viewValue) : cancel();
  }, [cancel, confirm, needConfirm, showTime, viewValue]));
  const [localMode, setLocalMode] = (0, _useSyncState.default)(mode);
  const triggerElementEnable = Boolean((0, _util.isFunction)(triggerElement) || triggerElement);
  const showAction = triggerElementEnable ? ['click'] : [];
  let triggerElementNode;
  if ((0, _util.isFunction)(triggerElement)) {
    triggerElementNode = triggerElement({
      text,
      hoverText,
      open,
      value
    });
  } else {
    triggerElementNode = triggerElement;
  }
  const handleOnSelect = (0, _react.useCallback)(newValue => {
    onSelect === null || onSelect === void 0 ? void 0 : onSelect(newValue);
    if (!needConfirm && !showTime) {
      confirm(newValue);
    } else {
      setViewValue(newValue);
    }
  }, [onSelect, needConfirm, showTime, confirm, setViewValue]);
  return /*#__PURE__*/_react.default.createElement(Popper, {
    destroyOnHide: true,
    popupStyle: popupStyle,
    getPopupContainer: triggerNode => {
      const _getPopupContainer = getPopupContainer !== null && getPopupContainer !== void 0 ? getPopupContainer : ctxGetPopupContainer;
      return _getPopupContainer(triggerNode);
    },
    visible: open,
    action: undefined,
    showAction: showAction,
    hideAction: [],
    popupClassName: (0, _clsx.default)(dropdownClassName, `${prefixCls}-dropdown`),
    disabled: disabled,
    onVisibleChange: setOpened,
    popup: /*#__PURE__*/_react.default.createElement(_List.ListBox, {
      className: `${prefixCls}__list__box`,
      onMouseDown: e => e.preventDefault(),
      ref: panelDivRef
    }, /*#__PURE__*/_react.default.createElement(_Context.DatePickerContextProvider, {
      value: {
        mode: localMode,
        setMode: setLocalMode,
        value,
        setValue: function setDateValue(v) {
          let autoCommit = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : !needConfirm;
          if (autoCommit) {
            confirm(typeof v === 'function' ? v(viewValue) : v);
          } else {
            setViewValue(v);
          }
        },
        showToday,
        renderExtraFooter,
        setOpened,
        showTime,
        needConfirm,
        showNow
      }
    }, /*#__PURE__*/_react.default.createElement(_picker.default, {
      defaultViewDate: defaultPickerValue,
      disabledDate: disabledDate,
      onDateMouseEnter: setHoverDate,
      onContentMouseLeave: () => setHoverDate(undefined),
      onSelect: handleOnSelect,
      value: viewValue,
      mode: mode,
      startOfWeek: startOfWeek,
      dateRender: dateRender,
      format: format,
      onCancel: cancel,
      onConfirm: () => confirm(viewValue),
      minDate: minDate,
      maxDate: maxDate,
      hideWeekNumbers: hideWeekNumbers,
      janX: janX
    }))),
    prefixCls: prefixCls,
    popupMotion: "slide-up",
    popupPlacement: direction === 'rtl' ? 'bottomRight' : 'bottomLeft',
    enableRestoreFocus: false,
    initialFocus: null,
    showArrow: false,
    arrowHasStroke: false,
    Transition: _SlideUpDownTransition.SlideUpDownTransition,
    enableMaxHeight: false,
    defaultMainAxisOffsetWithoutArrow: 4
  }, (() => {
    if (triggerElementNode) {
      // 如果用户传递了非 ReactElement 这里会直接报错，导致无法正常渲染
      return /*#__PURE__*/_react.default.cloneElement(triggerElementNode, {
        ref: (0, _composeRefs.default)(triggerElementRef, triggerElementNode.ref)
      });
    }
    return /*#__PURE__*/_react.default.createElement("div", {
      style: style,
      ref: inputDivRef,
      className: rootCls
    }, /*#__PURE__*/_react.default.createElement(_inputWrapper.default, {
      className: (0, _clsx.default)(className, `${prefixCls}-dateInput`),
      showClear: Boolean(viewValue),
      onClear: handleClear,
      prefixCls: prefixCls,
      onMouseUp: handleMouseUp,
      onMouseDown: onMouseDown,
      onClick: onClick,
      focused: focused,
      suffixIcon: suffixIcon,
      disabled: disabled,
      allowClear: allowClear
    }, /*#__PURE__*/_react.default.createElement(_Input.default, {
      id: id,
      size: (0, _ui.getInputSize)(format, _moment2.default),
      className: (0, _clsx.default)({
        [`${prefixCls}-input`]: true,
        [`${prefixCls}-input-placeholder`]: Boolean(hoverText)
      }),
      placeholder: placeholder,
      disabled: disabled,
      autoFocus: autoFocus,
      ref: inputRef,
      onKeyDown: handleKeyDown,
      onFocus: handleFocus,
      onBlur: handleBlur,
      onChange: handleInputChange,
      readOnly: inputReadOnly || (0, _util.isFunction)(format) && !(0, _util.isFunction)(parse),
      value: hoverText || text
    })));
  })());
}
