// @byted-service/bytedtrace/lib/sampler/condition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConditionFilter = exports.ConditionFilters = exports.ConditionSampler = void 0;
const span_1 = require("../span");
const DURATION_PARAM = '_duration';
const STATUS_CODE_PARAM = '_status_code';
const BUSINESS_STATUS_CODE_PARAM = '_biz_status_code';
const IS_ERROR_PARAM = '_is_error';
const EQUAL_SYMBOL = '=';
const NOT_EQUAL_SYMBOL = '!=';
const GREATER_THAN_SYMBOL = '>';
const GREATER_OR_EQUAL_SYMBOL = '>=';
const greaterOrEqualSymbolFunc = (left, right) => {
    return equalSymbolFunc(left, right) || greaterThanSymbolFunc(left, right);
};
const equalSymbolFunc = (left, right) => {
    return left === right;
};
const greaterThanSymbolFunc = (left, right) => {
    return compare(left, right, true);
};
const compare = (left, right, greater) => {
    if (typeof left === 'number' && typeof right === 'number') {
        return greater ? left > right : left < right;
    }
    /* istanbul ignore next */
    return false;
};
const notEqualSymbolFunc = (left, right) => {
    return left !== right;
};
const SymbolMap = {
    [EQUAL_SYMBOL]: equalSymbolFunc,
    [NOT_EQUAL_SYMBOL]: notEqualSymbolFunc,
    [GREATER_THAN_SYMBOL]: greaterThanSymbolFunc,
    [GREATER_OR_EQUAL_SYMBOL]: greaterOrEqualSymbolFunc,
};
class ConditionSampler {
    getFilters(spanType, name) {
        if (this.filterPack) {
            return this.filterPack.findConditionFilters(spanType, name);
        }
        return;
    }
    syncConfig(cfs = []) {
        this.filterPack = this.genFilters(cfs);
    }
    genFilters(cfs) {
        if (!cfs || cfs.length <= 0) {
            return;
        }
        const filterPack = new ConditionFilterPack();
        cfs.forEach((config) => {
            if (span_1.SamplingTraceType[config.samplingType]) {
                const conditionFilter = new ConditionFilter(config.samplingType, config.and);
                config.conditionMetas.forEach((meta) => {
                    const conditionMeta = this.genConditionMeta(meta);
                    if (conditionMeta) {
                        conditionFilter.append(conditionMeta);
                    }
                });
                filterPack.append(config.spanType, config.name, conditionFilter);
            }
        });
        return filterPack;
    }
    genConditionMeta(c) {
        let meta;
        const value = parseInt(c.value, 10);
        switch (c.param) {
            case DURATION_PARAM: {
                meta = new DurationConditionMeta(value);
                break;
            }
            case STATUS_CODE_PARAM: {
                meta = new StatusCodeConditionMeta(value);
                break;
            }
            case IS_ERROR_PARAM: {
                meta = new IsErrorConditionMeta(!!c.value);
                break;
            }
            case BUSINESS_STATUS_CODE_PARAM: {
                meta = new BusinessStatusCodeConditionMeta(value);
                break;
            }
        }
        if (!meta) {
            return;
        }
        if (SymbolMap[c.symbol]) {
            meta.setConditionSymbol(SymbolMap[c.symbol]);
        }
        else {
            return;
        }
        return meta;
    }
}
exports.ConditionSampler = ConditionSampler;
class ConditionFilterPack {
    constructor() {
        this.filters = {};
    }
    findConditionFilters(spanType, name) {
        const filter = this.filters[spanType];
        if (filter && filter[name]) {
            return filter[name];
        }
        return;
    }
    append(spanType, name, cf) {
        if (!this.filters[spanType]) {
            this.filters[spanType] = {};
        }
        this.appendByName(spanType, name, cf);
    }
    appendByName(spanType, name, cf) {
        const first = this.filters[spanType];
        let filters;
        if (first[name]) {
            filters = first[name];
        }
        else {
            filters = new ConditionFilters();
        }
        filters.filters.push(cf);
        first[name] = filters;
    }
}
class ConditionFilters {
    constructor() {
        this.filters = [];
    }
    sample(sp) {
        this.filters.forEach((filter) => {
            if (filter.filter(sp)) {
                return;
            }
        });
    }
}
exports.ConditionFilters = ConditionFilters;
class ConditionMeta {
    /* istanbul ignore next */
    filter(sp) {
        return false;
    }
    setConditionSymbol(conditionSymbol) {
        this.conditionSymbol = conditionSymbol;
    }
}
class ConditionFilter {
    constructor(samplingType, and) {
        this.conditions = [];
        this.samplingType = samplingType;
        this.and = and;
    }
    append(meta) {
        this.conditions.push(meta);
    }
    filter(sp) {
        if (!this.and) {
            for (let i = 0; i < this.conditions.length; i++) {
                const meta = this.conditions[i];
                if (meta.filter(sp)) {
                    sp.samplingTrace({ samplingTraceType: this.samplingType });
                    return true;
                }
            }
            /* istanbul ignore next */
            return false;
        }
        else {
            for (let i = 0; i < this.conditions.length; i++) {
                const meta = this.conditions[i];
                if (!meta.filter(sp)) {
                    /* istanbul ignore next */
                    return false;
                }
            }
            sp.samplingTrace({ samplingTraceType: this.samplingType });
            return true;
        }
    }
}
exports.ConditionFilter = ConditionFilter;
class DurationConditionMeta extends ConditionMeta {
    constructor(durationUs) {
        super();
        this.durationUs = durationUs;
    }
    filter(sp) {
        return this.conditionSymbol(sp.getDurationCost() / 1000, this.durationUs);
    }
}
class StatusCodeConditionMeta extends ConditionMeta {
    constructor(statusCode) {
        super();
        this.statusCode = statusCode;
    }
    filter(sp) {
        return this.conditionSymbol(sp.getStatusCode(), this.statusCode);
    }
}
class BusinessStatusCodeConditionMeta extends ConditionMeta {
    constructor(businessStatusCode) {
        super();
        this.businessStatusCode = businessStatusCode;
    }
    filter(sp) {
        return this.conditionSymbol(sp.getBusinessStatusCode(), this.businessStatusCode);
    }
}
class IsErrorConditionMeta extends ConditionMeta {
    constructor(isError) {
        super();
        this.isError = isError;
    }
    filter(sp) {
        return this.conditionSymbol(sp.getIsError(), this.isError);
    }
}
