// @byted-service/bytedtrace/lib/sampler/counter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WindowCounter = void 0;
class WindowCounter {
    constructor(bucketNum, maxTokenNum) {
        this.count = 0;
        this.oldIndex = 0;
        /**
         * 当前存储桶在存储桶数组中的下标
         */
        this.nowIndex = 0;
        this.tokenNum = 0;
        this.buckets = Array(Number(bucketNum)).fill(0);
        this.bucketNum = bucketNum;
        this.inWindow = 1;
        this.maxTokenNum = maxTokenNum;
    }
    /**
     * 打一次点
     * @param n
     */
    add(n) {
        const index = Number(this.nowIndex);
        this.buckets[index] += n;
        this.count += n;
        this.charge();
    }
    /**
     * 获取打点量
     * @returns
     */
    get() {
        return this.count;
    }
    /**
     * 首先获取打点量再打点
     * @param n
     * @returns
     */
    getAndAdd(n) {
        const count = this.get();
        this.add(n);
        return count;
    }
    zero() {
        this.count = 0;
    }
    tick() {
        this.cut();
        if (this.inWindow === this.bucketNum) {
            const oldIndex = Number(this.oldIndex);
            const bucket = this.buckets[oldIndex];
            this.count -= bucket;
            this.buckets[oldIndex] -= bucket;
            this.oldIndex++;
            if (this.oldIndex >= this.bucketNum) {
                this.oldIndex = 0;
            }
        }
        else {
            this.inWindow++;
        }
        this.incNowIndex();
    }
    incNowIndex() {
        if (this.nowIndex === this.bucketNum - 1) {
            this.nowIndex = 0;
        }
        else {
            this.nowIndex += 1;
        }
    }
    charge() {
        this.tokenNum = this.maxTokenNum;
    }
    cut() {
        if (this.tokenNum > -1) {
            this.tokenNum -= 1;
        }
    }
    /**
     * 返回是否过期
     * @returns
     */
    shouldExpired() {
        return this.tokenNum < 0;
    }
}
exports.WindowCounter = WindowCounter;
