// @byted-service/bytedtrace/lib/sampler/lowerBoud.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NoopLowerBoundSampler = exports.LowerBoundSampler = exports.minExpireTimeSecond = void 0;
const tslib_1 = require("tslib");
const status_1 = require("../status");
const seed_random_1 = tslib_1.__importDefault(require("seed-random"));
const env_1 = require("@byted-service/env");
exports.minExpireTimeSecond = 60;
const safeIntervalPercent = 0.8;
class LowerBoundSampler {
    constructor(minExpireTimeSecond, statCounter) {
        this.methodSet = new Set();
        // 这里跟 Golang 不一样
        this.random = (0, seed_random_1.default)((0, env_1.getIP)());
        this.statCounter = statCounter;
        this.minExpireTimeSecond = minExpireTimeSecond;
    }
    sampled(method) {
        if (!this.methodSet.has(method)) {
            if (this.methodSet.size >= this.maxMethodCount) {
                this.statCounter.recordWarningWithRateLimit(false, status_1.ErrLowerBoundSamplerMethodNumOverflow, '[bytedtrace] lower bound sampler method num overflow!');
                return false;
            }
            this.methodSet.add(method);
            return true;
        }
        else {
            return false;
        }
    }
    start() {
        if (this.intervalId) {
            return;
        }
        this.intervalId = setInterval(() => {
            /* istanbul ignore next */
            this.methodSet.clear();
        }, this.genExpireTime() * 1000).unref();
    }
    close() {
        clearInterval(this.intervalId);
    }
    syncConfig(config) {
        this.setExpireTime(config.lowerBoundSamplerExpireSecond);
        this.maxMethodCount = config.samplerMaxOperationCount;
    }
    genExpireTime() {
        const safeTimeSecond = safeIntervalPercent * this.expireSecond;
        const randomDuration = this.expireSecond - safeTimeSecond;
        let rt = 0;
        if (randomDuration > 0) {
            rt = this.random() * randomDuration;
        }
        return safeTimeSecond + rt;
    }
    setExpireTime(expireSecond) {
        if (expireSecond < this.minExpireTimeSecond) {
            expireSecond = this.minExpireTimeSecond;
        }
        const old = this.expireSecond;
        if (old !== expireSecond) {
            this.expireSecond = expireSecond;
            this.resetTimer();
        }
    }
    resetTimer() {
        clearInterval(this.intervalId);
        this.start();
    }
}
exports.LowerBoundSampler = LowerBoundSampler;
/* istanbul ignore next */
class NoopLowerBoundSampler {
    sampled() {
        return false;
    }
    start() {
        // do nothing
    }
    close() {
        // do nothing
    }
    syncConfig() {
        // do nothing
    }
}
exports.NoopLowerBoundSampler = NoopLowerBoundSampler;
