// @byted-service/bytedtrace/lib/sampler/sampler.js
"use strict";
/* tslint:disable max-classes-per-file */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AdaptiveSampler = exports.ProbabilitySampler = exports.PackSampler = exports.ByteSampler = void 0;
const tslib_1 = require("tslib");
const gConsts = tslib_1.__importStar(require("../consts"));
const gUtils = tslib_1.__importStar(require("../utils"));
const consts = tslib_1.__importStar(require("./consts"));
const confProvider = tslib_1.__importStar(require("../confProvider"));
const qpsCatcher_1 = require("./qpsCatcher");
const env_1 = require("@byted-service/env");
const lowerBoud_1 = require("./lowerBoud");
const MAX_INT64 = BigInt('9223372036854775807');
class ByteSampler {
    constructor(cf, env, statCounter) {
        this.env = env;
        this.statCounter = statCounter;
        this.sampler = new PackSampler(cf, env);
        this.initLowerBoundSampler(cf);
    }
    initLowerBoundSampler(cf) {
        this.lowerBoundSampler =
            cf.lowerBoundSamplerExpireSecond <= 0
                ? new lowerBoud_1.NoopLowerBoundSampler()
                : this.createLowerBoundSampler(cf);
    }
    updateLowerBoundSampler(cf) {
        const oldLowerBoundSampler = this.lowerBoundSampler;
        let newLowerBoundSampler = oldLowerBoundSampler;
        if (cf.lowerBoundSamplerExpireSecond > 0) {
            newLowerBoundSampler =
                newLowerBoundSampler instanceof lowerBoud_1.NoopLowerBoundSampler
                    ? this.createLowerBoundSampler(cf)
                    : newLowerBoundSampler;
        }
        else {
            /* istanbul ignore next */
            newLowerBoundSampler = new lowerBoud_1.NoopLowerBoundSampler();
        }
        newLowerBoundSampler.syncConfig(cf);
        this.lowerBoundSampler = newLowerBoundSampler;
        /* istanbul ignore next */
        if (newLowerBoundSampler !== oldLowerBoundSampler) {
            oldLowerBoundSampler.close();
        }
    }
    createLowerBoundSampler(cf) {
        const lowerBoundSampler = new lowerBoud_1.LowerBoundSampler(lowerBoud_1.minExpireTimeSecond, this.statCounter);
        lowerBoundSampler.syncConfig(cf);
        lowerBoundSampler.start();
        return lowerBoundSampler;
    }
    isSampled(base, operationName) {
        const res = this.sampler.isSampled(base, operationName);
        if (!res.isSampled) {
            if (this.lowerBoundSampler.sampled(operationName)) {
                res.isSampled = true;
                res.samplerType = gConsts.LOWER_BOUND_SAMPLER;
            }
        }
        // for benchmark
        res.isSampled = gConsts.IS_BENCHMARK ? true : res.isSampled;
        return res;
    }
    incAndGetSamplingRate(operationName) {
        return this.sampler.incAndGetSamplingRate(operationName);
    }
    close() {
        this.sampler.close();
        this.lowerBoundSampler.close();
    }
    syncConfig(config) {
        const old = this.sampler;
        this.sampler = new PackSampler(config, this.env);
        if (old) {
            old.close();
        }
        this.updateLowerBoundSampler(config);
    }
}
exports.ByteSampler = ByteSampler;
// 多个采样器集合.
class PackSampler {
    constructor(cf, env) {
        this.env = env;
        this.config = cf || {};
        this.useDefaultSampler = true;
        this.samplers = new Map();
        this.initConfig();
    }
    // 返回是否采样,采样器类型,采样概率.
    isSampled(base, operationName) {
        return this.getSampler(operationName).isSampled(base, operationName);
    }
    getSampler(operationName) {
        if (this.useDefaultSampler) {
            return this.defaultSampler;
        }
        return this.samplers.get(operationName) || this.defaultSampler;
    }
    incAndGetSamplingRate(operationName) {
        return this.getSampler(operationName).incAndGetSamplingRate(operationName);
    }
    close() {
        if (this.qpsCatcher) {
            this.qpsCatcher.close();
        }
    }
    syncConfig() {
        // do nothing
    }
    // Attention: this method is used when packSampler is init, so unlock.
    initConfig() {
        let maxOperationNum;
        maxOperationNum =
            typeof this.config.samplerMaxOperationCount === 'number'
                ? this.config.samplerMaxOperationCount
                : consts.MAX_OPERATION_FOR_SAMPLER;
        const qpsCaptureOption = {
            maxOperationNum,
            windowDuringMs: consts.DEFAULT_WINDOW_DURING_MS,
            windowNum: consts.DEFAULT_WINDOW_NUM,
            windowExpireTimeMs: consts.DEFAULT_WINDOW_EXPIRED_TIME_MS,
        };
        this.qpsCatcher = new qpsCatcher_1.QpsCatcher(false, qpsCaptureOption);
        this.defaultSampler = genSampler(this.config.defaultStrategy, this.qpsCatcher);
        const { detailStrategies } = this.config;
        if (!Array.isArray(detailStrategies)) {
            /* istanbul ignore next */
            return;
        }
        this.useDefaultSampler = false;
        let samplerCount = 0;
        for (const detail of detailStrategies) {
            if (gUtils.contain(detail.condition.clusters, (0, env_1.getCluster)()) &&
                gUtils.contain(detail.condition.idcs, (0, env_1.getIDC)()) &&
                gUtils.contain(detail.condition.stages, (0, env_1.getStage)())) {
                for (const method of detail.condition.methods) {
                    if (samplerCount > this.config.samplerMaxOperationCount) {
                        /* istanbul ignore next */
                        return;
                    }
                    if (!this.samplers.get(method)) {
                        this.samplers.set(method, genSampler({
                            samplingMode: detail.strategy.sampling_mode,
                            samplingRateProbability: detail.strategy
                                .sampling_rate_probability,
                            samplingCountPerMinAdaptive: detail.strategy
                                .sampling_count_per_min_adaptive,
                        }, this.qpsCatcher));
                        samplerCount++;
                    }
                }
            }
        }
    }
}
exports.PackSampler = PackSampler;
function genSampler(config, qpsCatcher) {
    if (config.samplingMode === confProvider.SAMPLING_MODE_PROBABILITY) {
        return new ProbabilitySampler(config.samplingRateProbability);
    }
    else {
        return new AdaptiveSampler(qpsCatcher, config.samplingCountPerMinAdaptive);
    }
}
// 概率采样器.
class ProbabilitySampler {
    constructor(samplingRate) {
        this.init(samplingRate);
    }
    init(samplingRate) {
        if (samplingRate > 1) {
            samplingRate = 1;
        }
        else if (samplingRate < 0) {
            samplingRate = 0;
        }
        const len = 10 ** (samplingRate + '').length;
        this.samplingBoundary =
            (MAX_INT64 * BigInt(samplingRate * len)) / BigInt(len);
        this.samplingRate = samplingRate;
    }
    isSampled(base) {
        return {
            isSampled: this.samplingBoundary >= base,
            samplerType: gConsts.PROBABILITY_SAMPLER,
            samplingRate: this.samplingRate,
        };
    }
    // Do nothing.
    incAndGetSamplingRate() {
        return this.samplingRate;
    }
    // No close.
    close() {
        // do nothing
    }
    syncConfig(config) {
        this.init(config.samplingRateProbability);
    }
}
exports.ProbabilitySampler = ProbabilitySampler;
// 自适应采样器.
class AdaptiveSampler {
    constructor(qpsCatcher, hitPerMinute) {
        this.qpsCatcher = qpsCatcher;
        this.hitPerMinute = hitPerMinute;
    }
    isSampled(base, operationName) {
        let qps;
        try {
            qps = this.qpsCatcher.incAndGetQps(operationName, 1);
        }
        catch (e) /* istanbul ignore next */ {
            return {
                isSampled: false,
                samplerType: gConsts.ADAPTIVE_SAMPLER,
                samplingRate: 0,
            };
        }
        if (qps === 0) {
            return {
                isSampled: true,
                samplerType: gConsts.ADAPTIVE_SAMPLER,
                samplingRate: 1,
            };
        }
        const samplingRate = this.hitPerMinute / (60 * qps);
        if (samplingRate > 1) {
            return {
                isSampled: true,
                samplerType: gConsts.ADAPTIVE_SAMPLER,
                samplingRate: 1,
            };
        }
        const len = 10 ** (samplingRate + '').length;
        const samplingBoundary = (MAX_INT64 * BigInt(samplingRate * len + '')) / BigInt(len);
        return {
            isSampled: samplingBoundary >= base,
            samplerType: gConsts.ADAPTIVE_SAMPLER,
            samplingRate,
        };
    }
    incAndGetSamplingRate(operationName) {
        const qps = this.qpsCatcher.incAndGetQps(operationName, 1);
        if (qps === -1) {
            return -1;
        }
        if (qps === 0) {
            return 1;
        }
        return (parseFloat(this.hitPerMinute) / 60) * qps;
    }
    close() {
        // do nothing
    }
    syncConfig() {
        // do nothing
    }
}
exports.AdaptiveSampler = AdaptiveSampler;
