// @universe-design/react/lib/config-provider/hooks/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getCssScopeFormatter = getCssScopeFormatter;
exports.useClassPrefix = useClassPrefix;
exports.useComponentSize = useComponentSize;
exports.useCssScope = useCssScope;
exports.useDirection = useDirection;
exports.useGetTargetContainer = useGetTargetContainer;
exports.useInputGroupSize = useInputGroupSize;
exports.useLocale = useLocale;
exports.useLocaleReceiver = useLocaleReceiver;
exports.useTheme = useTheme;
var _react = _interopRequireDefault(require("react"));
var _useContextSelector = require("use-context-selector");
var _ConfigContext = require("../ConfigContext");
var _size = require("../size");
function useComponentSize() {
  return (0, _useContextSelector.useContextSelector)(_ConfigContext.ConfigContext, o => o.componentSize);
}
function useInputGroupSize() {
  return (0, _useContextSelector.useContext)(_size.InputGroupSizeContext);
}
function useClassPrefix() {
  return (0, _useContextSelector.useContextSelector)(_ConfigContext.ConfigContext, o => o.rootPrefixCls);
}
function getCssScopeFormatter(classPrefix, scope, cuzScope) {
  return function () {
    let suffix = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
    return cuzScope ? `${cuzScope}${suffix}` : `${classPrefix}__${scope}${suffix}`;
  };
}
function useCssScope(scope, cuzScope) {
  const classPrefix = useClassPrefix();
  return _react.default.useMemo(() => getCssScopeFormatter(classPrefix, scope, cuzScope), [classPrefix, cuzScope, scope]);
}
function useDirection() {
  return (0, _useContextSelector.useContextSelector)(_ConfigContext.ConfigContext, o => o.direction);
}
function useGetTargetContainer() {
  return (0, _useContextSelector.useContextSelector)(_ConfigContext.ConfigContext, o => o.getTargetContainer);
}
function useLocale() {
  return (0, _useContextSelector.useContextSelector)(_ConfigContext.ConfigContext, o => o.locale);
}
function useLocaleReceiver(componentName) {
  const contextLocale = useLocale();
  return _react.default.useMemo(() => contextLocale.components[componentName], [componentName, contextLocale.components]);
}
function useTheme() {
  return (0, _useContextSelector.useContextSelector)(_ConfigContext.ConfigContext, o => o.theme);
}
