// @universe-design/react/es/grid/row.js
import _extends from "@babel/runtime/helpers/esm/extends";
import * as React from 'react';
import clsx from 'clsx';
import RowContext from './context';
import { useCssScope } from '../config-provider/hooks';
import { responsiveArray, ResponsiveObserve } from '../util';
import './style';
/**
 * @public
 */

export const Row = /*#__PURE__*/React.forwardRef((props, ref) => {
  const {
    children,
    align,
    justify,
    gutter = 0,
    wrap,
    style,
    className,
    ...otherProps
  } = props;
  const cssScope = useCssScope('row');
  const classes = clsx(cssScope(), {
    [cssScope(`--align-${align}`)]: align,
    [cssScope(`--justify-${justify}`)]: justify,
    [cssScope('--no-wrap')]: wrap === false
  }, className);
  const [screens, setScreens] = React.useState({
    xs: true,
    sm: true,
    md: true,
    lg: true,
    xl: true,
    xxl: true
  });
  const gutterRef = React.useRef(gutter);
  React.useEffect(() => {
    const token = ResponsiveObserve.subscribe(screen => {
      const currentGutter = gutterRef.current || 0;
      if (!Array.isArray(currentGutter) && typeof currentGutter === 'object' || Array.isArray(currentGutter) && (typeof currentGutter[0] === 'object' || typeof currentGutter[1] === 'object')) {
        setScreens(screen);
      }
    });
    return () => ResponsiveObserve.unsubscribe(token);
  }, []);
  const gutters = (() => {
    const results = [0, 0];
    const normalizedGutter = Array.isArray(gutter) ? gutter : [gutter, 0];
    normalizedGutter.forEach((g, index) => {
      if (typeof g === 'object') {
        for (let i = 0; i < responsiveArray.length; i++) {
          const breakpoint = responsiveArray[i];
          if (screens[breakpoint] && g[breakpoint] !== undefined) {
            results[index] = g[breakpoint];
            break;
          }
        }
      } else {
        results[index] = g || 0;
      }
    });
    return results;
  })();
  const rowStyle = React.useMemo(() => {
    const result = {};
    const [left, top] = gutters;
    if (left) {
      result.marginLeft = -left / 2;
      result.marginRight = -left / 2;
    }
    if (top) {
      result.marginTop = -top / 2;
      result.marginBottom = top / 2;
    }
    return result;
  }, [gutters]);
  return /*#__PURE__*/React.createElement(RowContext.Provider, {
    value: {
      gutter: gutters,
      wrap
    }
  }, /*#__PURE__*/React.createElement("div", _extends({
    ref: ref,
    style: {
      ...rowStyle,
      ...style
    },
    className: classes
  }, otherProps), children));
});
if (process.env.NODE_ENV !== "production") {
  Row.displayName = 'Row';
}
