// @universe-design/react/src/menu/stories/homeless.stories.tsx
/** @jsxImportSource @emotion/react */
import React, { useState } from 'react';
import type { MenuInfo } from '@universe-design/react';
import { Button, InputNumber, Menu, Dropdown, useStyleSheetOnce, Switch } from '@universe-design/react';

// 需要手动引入组件的样式。
import MailOutlined from '@universe-design/icons-react/MailOutlined';
import AppstoreOutlined from '@universe-design/icons-react/AppstoreOutlined';
import SettingOutlined from '@universe-design/icons-react/SettingOutlined';
import SideFoldOutlined from '@universe-design/icons-react/SideFoldOutlined';
import SideExpandOutlined from '@universe-design/icons-react/SideExpandOutlined';
import RightBordersOutlined from '@universe-design/icons-react/RightBordersOutlined';
import LinkCopyOutlined from '@universe-design/icons-react/LinkCopyOutlined';
import ShareOutlined from '@universe-design/icons-react/ShareOutlined';
import WindowNewOutlined from '@universe-design/icons-react/WindowNewOutlined';
import { Global } from '@emotion/react';

export default {
  title: 'Components/Menu/临时示例', // 注意：这里需要修改成自己对应 Component name。
  component: Menu,
};

// eslint-disable-next-line max-lines-per-function
export const AllLayout = () => {
  const [current, setCurrent] = useState('');
  const [collapsed, setCollapsed] = useState(false);
  const [textOverflow, setTextOverflow] = useState(true);

  const toggleCollapsed = () => setCollapsed(!collapsed);
  const handleClick = (e: MenuInfo) => {
    console.log('click ', e);
    setCurrent(e.key as string);
  };

  const longText = new Array(10).fill('long').join('_');

  return (
    <>
      <Global
        styles={{
          body: {
            backgroundColor: '#F5F6F7',
          },
        }}
      />
      是否溢出收起: <Switch checked={textOverflow} onChange={setTextOverflow} />
      <div style={{ marginTop: 50 }}>
        <Menu mode="horizontal" defaultSelectedKeys={['2']} textOverflow={textOverflow}>
          <Menu.Item key="1">{longText}</Menu.Item>
          <Menu.Item key="2">nav 2</Menu.Item>
          <Menu.Item key="3">nav 3</Menu.Item>
        </Menu>
      </div>
      <div>
        <Menu onClick={handleClick} selectedKeys={[current]} mode="horizontal" textOverflow={textOverflow}>
          <Menu.Item key="mail" icon={<MailOutlined />}>
            Navigation One
          </Menu.Item>
          <Menu.Item key="app" disabled icon={<AppstoreOutlined />}>
            {longText}
          </Menu.Item>
          <Menu.SubMenu key="SubMenu" icon={<SettingOutlined />} title="Navigation Three - Submenu">
            <Menu.ItemGroup title="Item 1">
              <Menu.Item key="setting:1">Option 1</Menu.Item>
              <Menu.Item key="setting:2">Option 2</Menu.Item>
            </Menu.ItemGroup>
            <Menu.ItemGroup title="Item 2">
              <Menu.Item key="setting:3">Option 3</Menu.Item>
              <Menu.Item key="setting:4">Option 4</Menu.Item>
              <Menu.SubMenu title="Sub Item">
                <Menu.Item key="setting:5">Option 3</Menu.Item>
                <Menu.Item key="setting:6">Option 4</Menu.Item>
              </Menu.SubMenu>
            </Menu.ItemGroup>
          </Menu.SubMenu>
          <Menu.Item key="alipay">
            <a href="https://universe.bytedance.net" target="_blank" rel="noopener noreferrer">
              Navigation Four - Link
            </a>
          </Menu.Item>
        </Menu>
      </div>
      <div style={{ marginTop: 50 }}>
        <Menu
          onClick={handleClick}
          style={{ width: 240 }}
          defaultSelectedKeys={['a1']}
          defaultOpenKeys={['sub1']}
          mode="inline"
          textOverflow={textOverflow}
        >
          <Menu.SubMenu key="sub1" icon={<MailOutlined />} title="Navigation One">
            <Menu.ItemGroup key="g1" title="Item 1">
              <Menu.Item key="a1">Option 1</Menu.Item>
              <Menu.Item key="a2">Option 2</Menu.Item>
            </Menu.ItemGroup>
            <Menu.ItemGroup key="g2" title="Item 2">
              <Menu.Item key="3">Option 3</Menu.Item>
              <Menu.Item key="4">Option 4</Menu.Item>
            </Menu.ItemGroup>
          </Menu.SubMenu>
          <Menu.SubMenu key="sub2" icon={<AppstoreOutlined />} title={longText}>
            <Menu.Item key="5">{longText}</Menu.Item>
            <Menu.Item key="6">Option 6</Menu.Item>
            <Menu.SubMenu key="sub3" title={longText}>
              <Menu.Item key="7">Option 7</Menu.Item>
              <Menu.Item key="8">Option 8</Menu.Item>
            </Menu.SubMenu>
          </Menu.SubMenu>
          <Menu.SubMenu key="sub4" icon={<SettingOutlined />} title="Navigation Three">
            <Menu.Item key="9">Option 9</Menu.Item>
            <Menu.Item key="10">Option 10</Menu.Item>
            <Menu.Item key="11">Option 11</Menu.Item>
            <Menu.Item key="12">Option 12</Menu.Item>
          </Menu.SubMenu>
        </Menu>
      </div>
      <div style={{ width: 240, marginTop: 50 }}>
        <Button onClick={toggleCollapsed} style={{ marginBottom: 16 }}>
          {React.createElement(collapsed ? SideFoldOutlined : SideExpandOutlined)}
        </Button>
        <Menu
          defaultSelectedKeys={['1']}
          defaultOpenKeys={['sub1']}
          mode="inline"
          inlineCollapsed={collapsed}
          textOverflow={textOverflow}
        >
          <Menu.Item key="1">Option 1</Menu.Item>
          <Menu.Item key="2" icon={<AppstoreOutlined />}>
            Option 2
          </Menu.Item>
          <Menu.Item key="3" icon={<SettingOutlined />}>
            Option 3
          </Menu.Item>
          <Menu.SubMenu key="sub1" icon={<MailOutlined />} title="Navigation One A">
            <Menu.Item key="5" icon={<MailOutlined />}>
              Option 5
            </Menu.Item>
            <Menu.Item key="6">Option 6</Menu.Item>
            <Menu.Item key="7">Option 7</Menu.Item>
            <Menu.Item key="8">Option 8</Menu.Item>
          </Menu.SubMenu>
          <Menu.SubMenu key="sub2" icon={<AppstoreOutlined />} title="Navigation Two">
            <Menu.Item key="9">Option 9</Menu.Item>
            <Menu.Item key="10">Option 10</Menu.Item>
            <Menu.SubMenu key="sub3" title="Submenu">
              <Menu.Item key="11">Option 11</Menu.Item>
              <Menu.Item key="12">Option 12</Menu.Item>
            </Menu.SubMenu>
          </Menu.SubMenu>
        </Menu>
      </div>
      <div css={{ display: 'flex', marginTop: 50 }}>
        {[Menu].map((C, index) => (
          <div css={{ width: '50%' }} key={index}>
            <div css={{ width: 240 }}>
              <C
                onClick={handleClick}
                mode="vertical"
                inlineCollapsed={collapsed}
                expandIcon={<RightBordersOutlined />}
                textOverflow={textOverflow}
              >
                <C.Item key="0">Option 0</C.Item>
                <C.SubMenu key="sub1" icon={<MailOutlined />} title="Navigation One">
                  <C.ItemGroup title="Item 1">
                    <C.Item key="1">Option 1</C.Item>
                    <C.Item key="2">Option 2</C.Item>
                  </C.ItemGroup>
                  <C.ItemGroup title="Iteom 2">
                    <C.Item key="3">Option 3</C.Item>
                    <C.Item key="4">Option 4</C.Item>
                  </C.ItemGroup>
                </C.SubMenu>
                <C.SubMenu key="sub2" icon={<AppstoreOutlined />} title="Navigation Two">
                  <C.Item key="5">Option 5</C.Item>
                  <C.Item key="6">Option 6</C.Item>
                  <C.SubMenu key="sub3" title="Submenu">
                    <C.Item key="7">Option 7</C.Item>
                    <C.Item key="8">Option 8</C.Item>
                  </C.SubMenu>
                </C.SubMenu>
                <C.SubMenu key="sub4" icon={<SettingOutlined />} title="Navigation Three">
                  <C.Item key="9">Option 9</C.Item>
                  <C.Item key="10">Option 10</C.Item>
                  <C.Item key="11">Option 11</C.Item>
                  <C.Item key="12">Option 12</C.Item>
                </C.SubMenu>
              </C>
            </div>
          </div>
        ))}
      </div>
      <Menu style={{ width: 180 }} mode="normal">
        <Menu.Item icon={<WindowNewOutlined />}>{longText}</Menu.Item>
        <Menu.Item icon={<ShareOutlined />}>分享</Menu.Item>
        <Menu.Divider />
        <Menu.Item icon={<LinkCopyOutlined />}>复制链接</Menu.Item>
        <Menu.Divider />
        <Menu.Item style={{ textIndent: 24 }}>创建副本</Menu.Item>
        <Menu.Divider />
        <Menu.Item style={{ textIndent: 24 }}>添加至文件夹</Menu.Item>
        <Menu.Item style={{ textIndent: 24 }}>添加至快捷访问</Menu.Item>
        <Menu.Item style={{ textIndent: 24 }}>添加至收藏</Menu.Item>
        <Menu.SubMenu style={{ textIndent: 24 }} title={longText} key="group">
          <Menu.Item icon={<WindowNewOutlined />}>新建标签中打开</Menu.Item>
          <Menu.Item icon={<ShareOutlined />} key="share">
            分享
          </Menu.Item>
        </Menu.SubMenu>
      </Menu>
    </>
  );
};

export const CustomerInlineIndent = () => (
  <>
    <Menu
      style={{ width: 240 }}
      defaultSelectedKeys={['a1']}
      defaultOpenKeys={['sub1']}
      mode="inline"
      inlineIndent={18}
    >
      <Menu.SubMenu key="sub1" icon={<MailOutlined />} title="Navigation One">
        <Menu.Item key="a1">Option 1</Menu.Item>
        <Menu.Item key="a2">Option 2</Menu.Item>
        <Menu.Item key="3">Option 3</Menu.Item>
        <Menu.Item key="4">Option 4</Menu.Item>
      </Menu.SubMenu>
      <Menu.SubMenu key="sub2" icon={<AppstoreOutlined />} title="Navigation Two">
        <Menu.Item key="5">Option 5</Menu.Item>
        <Menu.Item key="6">Option 6</Menu.Item>
        <Menu.SubMenu key="sub3" title="Submenu" icon={<AppstoreOutlined />}>
          <Menu.Item key="7">Option 7</Menu.Item>
          <Menu.Item key="8">Option 8</Menu.Item>
        </Menu.SubMenu>
      </Menu.SubMenu>
      <Menu.SubMenu key="sub4" icon={<SettingOutlined />} title="Navigation Three">
        <Menu.Item key="9">Option 9</Menu.Item>
        <Menu.Item key="10">Option 10</Menu.Item>
        <Menu.Item key="11">Option 11</Menu.Item>
        <Menu.Item key="12">Option 12</Menu.Item>
      </Menu.SubMenu>
    </Menu>
    <Menu
      style={{ width: 240 }}
      defaultSelectedKeys={['a1']}
      defaultOpenKeys={['sub1']}
      mode="inline"
      inlineIndent={18}
    >
      <Menu.SubMenu key="sub1" title="Navigation One">
        <Menu.Item key="a1">Option 1</Menu.Item>
        <Menu.Item key="a2">Option 2</Menu.Item>
        <Menu.Item key="3">Option 3</Menu.Item>
        <Menu.Item key="4">Option 4</Menu.Item>
      </Menu.SubMenu>
      <Menu.SubMenu key="sub2" title="Navigation Two">
        <Menu.Item key="5">Option 5</Menu.Item>
        <Menu.Item key="6">Option 6</Menu.Item>
        <Menu.SubMenu key="sub3" title="Submenu" icon={<AppstoreOutlined />}>
          <Menu.Item key="7">Option 7</Menu.Item>
          <Menu.Item key="8">Option 8</Menu.Item>
        </Menu.SubMenu>
      </Menu.SubMenu>
      <Menu.SubMenu key="sub4" title="Navigation Three">
        <Menu.Item key="9">Option 9</Menu.Item>
        <Menu.Item key="10">Option 10</Menu.Item>
        <Menu.Item key="11">Option 11</Menu.Item>
        <Menu.Item key="12">Option 12</Menu.Item>
      </Menu.SubMenu>
    </Menu>
  </>
);

export const InputNumberInMenu = () => (
  <Menu style={{ width: 240 }} defaultOpenKeys={['8']} triggerSubMenuAction="click" mode="vertical" inlineIndent={18}>
    <Menu.SubMenu key="sub1" title="Navigation One">
      <Menu.Item key="a1">Option 1</Menu.Item>
      <Menu.Item key="a2">Option 2</Menu.Item>
      <Menu.Item key="3">Option 3</Menu.Item>
      <Menu.Item key="4">Option 4</Menu.Item>
    </Menu.SubMenu>
    <Menu.SubMenu key="sub2" title="Navigation Two">
      <Menu.Item key="5">Option 5</Menu.Item>
      <Menu.Item key="6">Option 6</Menu.Item>
      <Menu.SubMenu key="sub3" title="Submenu" icon={<AppstoreOutlined />}>
        <Menu.Item key="7">Option 7</Menu.Item>
        <Menu.Item key="8">
          <InputNumber onClick={e => e.stopPropagation()} style={{ marginTop: 6 }} />
        </Menu.Item>
      </Menu.SubMenu>
    </Menu.SubMenu>
    <Menu.SubMenu key="sub4" title="Navigation Three">
      <Menu.Item key="9">Option 9</Menu.Item>
      <Menu.Item key="10">Option 10</Menu.Item>
      <Menu.Item key="11">Option 11</Menu.Item>
      <Menu.Item key="12">Option 12</Menu.Item>
    </Menu.SubMenu>
  </Menu>
);
export const SmoothPointerMove = () => {
  useStyleSheetOnce(`
  .title, .description {
    margin: 0;
  }
  .smooth-pointer .story__menu-normal-item .description {
    display: none;
  }

  .story__menu-normal-item:hover .description {
    display: block;
  }

  `);

  const menu = (
    <Menu mode="normal" className="smooth-pointer" style={{ width: 300 }}>
      <Menu.Item key="9">
        <div>
          <p className="title">Title 1 </p>
          <p className="description">Description </p>
        </div>
      </Menu.Item>
      <Menu.Item key="10">
        <div>
          <p className="title">Title 2 </p>
          <p className="description">Description </p>
        </div>
      </Menu.Item>
      <Menu.Item key="11">
        <div>
          <p className="title">Title 3 </p>
          <p className="description">Description </p>
        </div>
      </Menu.Item>
      <Menu.Item key="12">
        <div>
          <p className="title">Title1 </p>
          <p className="description">Description </p>
        </div>
      </Menu.Item>
    </Menu>
  );

  return (
    <Dropdown overlay={menu}>
      <Button>Menu Trigger</Button>
    </Dropdown>
  );
};
