// @universe-design/react/src/menu/stories/screenshot.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C2CollapseScreenshot from '../demo/2.collapse.screenshot';
import CBaseScreenshot from '../demo/base.screenshot';
import CCheckerScreenshot from '../demo/checker.screenshot';
import CHorizonalScreenshot from '../demo/horizonal.screenshot';
import CInlineScreenshot from '../demo/inline.screenshot';
import CMenuLinkScreenshot from '../demo/menu-link.screenshot';
import CVerticalScreenshot from '../demo/vertical.screenshot';
if (!(C2CollapseScreenshot as Story).storyName) {
  (C2CollapseScreenshot as Story).storyName = '2 collapse screenshot';
}
if (!(CBaseScreenshot as Story).storyName) {
  (CBaseScreenshot as Story).storyName = 'Base screenshot';
}
if (!(CCheckerScreenshot as Story).storyName) {
  (CCheckerScreenshot as Story).storyName = 'Checker screenshot';
}
if (!(CHorizonalScreenshot as Story).storyName) {
  (CHorizonalScreenshot as Story).storyName = 'Horizonal screenshot';
}
if (!(CInlineScreenshot as Story).storyName) {
  (CInlineScreenshot as Story).storyName = 'Inline screenshot';
}
if (!(CMenuLinkScreenshot as Story).storyName) {
  (CMenuLinkScreenshot as Story).storyName = 'Menu link screenshot';
}
if (!(CVerticalScreenshot as Story).storyName) {
  (CVerticalScreenshot as Story).storyName = 'Vertical screenshot';
}

export default {
  title: 'Components/Menu/截图测试',
  tags: ['autodocs'],
};

export {
  C2CollapseScreenshot,
  CBaseScreenshot,
  CCheckerScreenshot,
  CHorizonalScreenshot,
  CInlineScreenshot,
  CMenuLinkScreenshot,
  CVerticalScreenshot,
};
