// @universe-design/react/src/menu/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Horizontal from '../demo/0.horizontal';
import C1Vertical from '../demo/1.vertical';
import C2Collapse from '../demo/2.collapse';
import C3PopSubMenu from '../demo/3.pop-sub-menu';
import C4ResidentSubmenu from '../demo/4.resident-submenu';
import C5ContextMenu from '../demo/5.context-menu';
if (!(C0Horizontal as Story).storyName) {
  (C0Horizontal as Story).storyName = '0. 顶部导航';
  (C0Horizontal as Story).parameters = {
    docs: {
      description: {
        story: '水平的顶部导航菜单。往往跟 [Layout](/develop/react/layout) 组件一起使用。\n\n',
      },
    },
  };
}
if (!(C1Vertical as Story).storyName) {
  (C1Vertical as Story).storyName = '1. 内嵌菜单';
  (C1Vertical as Story).parameters = {
    docs: {
      description: {
        story: '垂直菜单，子菜单内嵌在菜单区域。\n\n',
      },
    },
  };
}
if (!(C2Collapse as Story).storyName) {
  (C2Collapse as Story).storyName = '2. 缩起内嵌菜单';
  (C2Collapse as Story).parameters = {
    docs: {
      description: {
        story: '内嵌菜单可以被缩起/展开。你可以在 Layout 里查看侧边布局结合的完整示例。\n\n',
      },
    },
  };
}
if (!(C3PopSubMenu as Story).storyName) {
  (C3PopSubMenu as Story).storyName = '2. 弹出垂直菜单';
  (C3PopSubMenu as Story).parameters = {
    docs: {
      description: {
        story: '子菜单是弹出的形式。\n\n',
      },
    },
  };
}
if (!(C4ResidentSubmenu as Story).storyName) {
  (C4ResidentSubmenu as Story).storyName = '4. 子菜单常驻';
  (C4ResidentSubmenu as Story).parameters = {
    docs: {
      description: {
        story: '始终保持子菜单的展开。\n\n',
      },
    },
  };
}
if (!(C5ContextMenu as Story).storyName) {
  (C5ContextMenu as Story).storyName = '4. 右键菜单';
  (C5ContextMenu as Story).parameters = {
    docs: {
      description: {
        story: '也可以模拟出右键菜单。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Menu/官网示例',
  tags: ['autodocs'],
};

export { C0Horizontal, C1Vertical, C2Collapse, C3PopSubMenu, C4ResidentSubmenu, C5ContextMenu };
