// @universe-design/react/lib/color-picker/pickers/ColorPicker.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ColorPicker = void 0;
var _react = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _type = require("../type");
var _util = require("../util");
var _RightBoldOutlined = _interopRequireDefault(require("@universe-design/icons-react/RightBoldOutlined"));
var _hooks = require("../../config-provider/hooks");
var _colors = require("../colors");
var _Palette = require("./Palette");
var _const = require("../const");
var _useComplementEvent = require("../../hooks/useComplementEvent");
var _ColorPickerContext = require("../ColorPickerContext");
var _ColorPickerColorBlock = require("../components/ColorPickerColorBlock");
require("../style");
/**
 * @public
 * @docgen
 */

const Blocks = props => {
  const {
    colorGroup
  } = props;
  const {
    rootCssScope
  } = (0, _ColorPickerContext.useColorPickerContext)();
  const prefix = rootCssScope('-swatches');
  const isEmptyColorGroup = (0, _react.useMemo)(() => (0, _util.judgeEmptyColorGroup)(colorGroup), [colorGroup]);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(prefix, isEmptyColorGroup && `${prefix}-empty`)
  }, colorGroup.map((colors, index) => /*#__PURE__*/_react.default.createElement("div", {
    key: index,
    className: `${prefix}-row`
  }, colors.map(color => /*#__PURE__*/_react.default.createElement(_ColorPickerColorBlock.ColorPickerColorBlock, {
    key: JSON.stringify(color),
    color: color
  })))));
};
const ColorPicker = props => {
  const locale = (0, _hooks.useLocale)();
  const {
    colorGroup = _colors.DEFAULT_COLOR_GROUP,
    showMoreColor = true,
    moreBtnText = locale.components.colorPicker.moreColor,
    className,
    style,
    onSaturationDragEnd,
    onHueDragEnd
  } = props;
  const [currentStep, setCurrentStep] = (0, _react.useState)(_type.ColorPickerStep.inHome);
  const colorPickerCtx = (0, _ColorPickerContext.useColorPicker)(props);
  const {
    rootCssScope
  } = colorPickerCtx;
  const prefix = rootCssScope();
  const isEmptyColorGroup = (0, _react.useMemo)(() => (0, _util.judgeEmptyColorGroup)(colorGroup), [colorGroup]);
  const containerRef = (0, _react.useRef)(null);
  const moreColorBtnRef = (0, _react.useRef)(null);
  (0, _useComplementEvent.useComplementEvent)('pointerdown', {
    listener: _react.default.useCallback(() => {
      setCurrentStep(_type.ColorPickerStep.inHome);
    }, []),
    // 这里必须有两个 ref，是因为点击 moreColorBtn 的时候会跳转到另一页，在事件触发的过程中，moreColorBtn 已经消失了，所以会判断为点击了外面，这个写法是为了保证事件从 moreColor 出来也可以判断是在组件内部
    elements: _react.default.useMemo(() => [containerRef, moreColorBtnRef.current], []),
    skipped: currentStep !== _type.ColorPickerStep.inMoreColor
  });
  const renderMoreColorBtn = () => {
    if (showMoreColor === false) {
      return null;
    }
    return /*#__PURE__*/_react.default.createElement("div", {
      ref: moreColorBtnRef,
      className: (0, _clsx.default)(`${prefix}-more`, isEmptyColorGroup && `${prefix}-more-empty`),
      onClick: () => setCurrentStep(_type.ColorPickerStep.inMoreColor)
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: `${prefix}-more-left`
    }, /*#__PURE__*/_react.default.createElement("img", {
      src: _const.MOREBTNICON
    }), /*#__PURE__*/_react.default.createElement("span", null, moreBtnText)), /*#__PURE__*/_react.default.createElement(_RightBoldOutlined.default, null));
  };
  return /*#__PURE__*/_react.default.createElement(_ColorPickerContext.ColorPickerProvider, {
    value: colorPickerCtx
  }, (() => {
    if (currentStep === _type.ColorPickerStep.inHome) {
      return /*#__PURE__*/_react.default.createElement("div", {
        ref: containerRef,
        className: (0, _clsx.default)(prefix, className),
        style: style
      }, /*#__PURE__*/_react.default.createElement(Blocks, {
        colorGroup: colorGroup
      }), renderMoreColorBtn());
    }
    return /*#__PURE__*/_react.default.createElement(_Palette.Palette, {
      onSaturationDragEnd: onSaturationDragEnd,
      onHueDragEnd: onHueDragEnd,
      ref: containerRef,
      className: className,
      style: style
    });
  })());
};
exports.ColorPicker = ColorPicker;
