// @universe-design/react/lib/color-picker/pickers/SimpleBlocksColorPicker.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SimpleBlocksColorPicker = void 0;
var _react = _interopRequireDefault(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _indexLight = require("@universe-design/styles/cjs/universe-design/index.light.atom");
var _ColorPickerContext = require("../ColorPickerContext");
var _space = require("../../space");
var _ColorPickerColorBlock = require("../components/ColorPickerColorBlock");
require("../style");
/**
 * @public
 * `<SimpleBlocksColorPicker />` 组件的属性。
 */

const defaults = {
  colorGroup: [[undefined, _indexLight.udtoken_colorpicker_red, _indexLight.udtoken_colorpicker_orange, _indexLight.udtoken_colorpicker_yellow, _indexLight.udtoken_colorpicker_green, _indexLight.udtoken_colorpicker_turquoise], [_indexLight.udtoken_colorpicker_blue, _indexLight.udtoken_colorpicker_wathet, _indexLight.udtoken_colorpicker_indigo, _indexLight.udtoken_colorpicker_purple, _indexLight.udtoken_colorpicker_violet, _indexLight.udtoken_colorpicker_neutral]]
};
const SimpleBlocksColorPicker = props => {
  const {
    className,
    style,
    colorGroup = defaults.colorGroup
  } = props;
  const colorPickerCtx = (0, _ColorPickerContext.useColorPicker)(props);
  const {
    rootCssScope
  } = colorPickerCtx;
  return /*#__PURE__*/_react.default.createElement(_ColorPickerContext.ColorPickerProvider, {
    value: colorPickerCtx
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(rootCssScope(), rootCssScope('-simple'), className),
    style: style
  }, /*#__PURE__*/_react.default.createElement(_space.Space, {
    direction: "vertical"
  }, colorGroup.map((row, i) => /*#__PURE__*/_react.default.createElement(_space.Space, {
    key: i,
    inline: false
  }, row.map(color => /*#__PURE__*/_react.default.createElement(_ColorPickerColorBlock.ColorPickerColorBlock, {
    key: typeof color === 'undefined' ? 'undefined' : JSON.stringify(color),
    color: color
  })))))));
};
exports.SimpleBlocksColorPicker = SimpleBlocksColorPicker;
