// @universe-design/react/src/notification/index.tsx
import React, { useRef } from 'react';
import ReactDOM from 'react-dom';
import { CrossRootConfigProvider } from '../config-provider';
import type { NotificationProps } from './Notification';
import Notification, { createNotificationWrapper } from './Notification';

import './style';

let globalNotification: Notification | undefined;

function getDefaultContainer() {
  const div = document.createElement('div');
  document.body.appendChild(div);
  return div;
}

function singletonRef(elem: Notification | null) {
  if (elem) {
    globalNotification = elem;
  }
}

/**
 * 用于多次重复 renderGlobalToast 和销毁自动生成的 DOM 元素
 */
class NotificationWrapper extends React.Component<NotificationProps> {
  private div?: HTMLElement;
  private isDefaultContainer = true;

  componentWillUnmount() {
    globalNotification = undefined;
    if (this.isDefaultContainer && this.div) {
      this.div.remove();
      this.div = undefined;
    }
  }

  getContainer = () => {
    const { getContainer } = this.props;
    if (!this.div) {
      const c = getContainer?.();
      this.div = c || getDefaultContainer();
      if (c) {
        this.isDefaultContainer = false;
      }
    }
    return this.div;
  };

  render() {
    return <Notification {...this.props} ref={singletonRef} getContainer={this.getContainer} />;
  }
}

export function getGlobalNotification(): Promise<Notification> {
  if (globalNotification) {
    return Promise.resolve(globalNotification);
  }
  return new Promise(resolve => {
    const div = getDefaultContainer();
    // 如果没有使用 useToast 或者 renderGlobalToast 初始化 toast，
    // 则默认取第一个 ConfigProvider 提供的配置
    ReactDOM.render(
      <CrossRootConfigProvider>
        <NotificationWrapper
          ref={() => {
            if (globalNotification) {
              resolve(globalNotification);
            }
          }}
          getContainer={() => div}
        />
      </CrossRootConfigProvider>,
      div,
    );
  });
}

export const notification = createNotificationWrapper(getGlobalNotification);
export { Notification };

/**
 * 用于创建可以获取 ConfigContext 的全局 Notification 单例
 * @example
 * ```tsx
 * <ConfigProvider>
 *   {renderGlobalNotification({ className: 'my-toast' }, document.getElementById('my-container'))}
 *   <App />
 * </ConfigProvider>
 * ```
 *
 * @param props: ToastProps
 * @param container?: HTMLElement
 * @returns React.ReactPortal
 */
export function renderGlobalNotification(props?: NotificationProps) {
  return <NotificationWrapper {...props} />;
}

/**
 * 获取一个局部的
 * @returns
 */
export function useNotification(props?: NotificationProps, container?: HTMLElement) {
  const ref = useRef<Notification>(null);
  const mountDivRef = useRef<HTMLDivElement>();
  const notificationIns = createNotificationWrapper(() => {
    if (ref.current) {
      return Promise.resolve(ref.current);
    }
    return getGlobalNotification();
  });
  const _getDefaultContainer = () => {
    if (mountDivRef.current) {
      return mountDivRef.current;
    }
    mountDivRef.current = getDefaultContainer();
    return mountDivRef.current;
  };
  return [
    notificationIns,
    <Notification
      key="use-notification"
      {...props}
      ref={ref}
      getContainer={() => container || _getDefaultContainer()}
    />,
  ] as const;
}

export type {
  NotificationInstanceProps,
  NotificationConfigOrContentType,
  NotificationConfigProps,
  NotificationInstanceMethod,
} from './Notification';
