// @universe-design/react/src/notification/manager.tsx
import type { CSSProperties, FC } from 'react';
import React, { createRef, useEffect, useRef } from 'react';
import ReactDOM from 'react-dom';
import clsx from 'clsx';
import { CSSTransition } from 'react-transition-group';
import type { CSSTransitionProps } from 'react-transition-group/CSSTransition';
import type { TransitionStatus } from 'react-transition-group/Transition';
import { useContextSelector } from 'use-context-selector';
import merge from 'lodash/merge';
import type { ConsumerProps } from '../config-provider';
import { ConfigContext } from '../config-provider';
import type { ConfigProviderProps } from '../config-provider/interface';
import { getCssScopeFormatter, useClassPrefix } from '../config-provider/hooks';
import TransitionGroup from './TransitionGroup';
import { useEventCallback } from '../hooks';

const DEFAULT_CLOSE_DURATION = 3000;
const DEFAULT_TRANSITION_DURATION = 300;
const DEFAULT_PLACEMENT = 'topRight';
export const BASE_MESSAGE_LENGTH = 20;

let messageIdentity = 0;

function getIdentity() {
  messageIdentity += 1;
  return `ud-msg-manager-${messageIdentity}`;
}

interface CustomToastConsumerProps {
  managerGlobalRootPrefixCls: string;
  managerGlobalToast: ConfigProviderProps['toast'];
}

const ManagerWrapConsumer: FC<ConsumerProps<CustomToastConsumerProps>> = props => {
  const managerGlobalRootPrefixCls = useClassPrefix();
  const managerGlobalToast = useContextSelector(ConfigContext, v => v.toast);
  return props.children({
    managerGlobalRootPrefixCls,
    managerGlobalToast,
  });
};

/**
 * React.Children.map 会对 jsx 的 key 属性进行转码，参考：https://github.com/facebook/react/blob/c1a53ad2b2f29f30412c682b42bdd6ac3227eafa/packages/react/src/ReactChildren.js#L63
 * bota 不能支持 unicode 的转码，所以需要处理一下，参考 https://stackoverflow.com/a/30106551/3053930
 */
function getEscapedKey(key: React.Key): React.Key {
  if (typeof key !== 'string') {
    return key;
  }
  const codeUnits = new Uint16Array(key.length);
  for (let i = 0; i < codeUnits.length; i++) {
    codeUnits[i] = key.charCodeAt(i);
  }
  return btoa(String.fromCharCode(...new Uint8Array(codeUnits.buffer))).replace(/=/g, '-');
}
interface RenderMessageArgs {
  toastConfigCls?: string;
  transitionStatus: TransitionStatus;
  rootPrefixCls: string;
  close: () => void;
  startTimer: () => void;
  stopTimer: () => void;
}

export type MessagePlacement = 'topLeft' | 'topRight' | 'bottomLeft' | 'bottomRight' | 'top';

export interface NotificationManagerProps {
  style?: CSSProperties;
  prefixCls?: string;
  className?: string;
  /**
   * 持续显示的时长，传 null 和 undefined 不自动关闭, 单位毫秒
   */
  duration?: number | null;
  /**
   * CSSTransition 的 props
   */
  transitionProps?: CSSTransitionProps;
  /**
   * 最多显示多少个消息，超过的消息将在队列中维护
   */
  maxCount?: number;
  /**
   * 距离顶部的距离
   */
  top?: number;
  /**
   * 距离底部的距离
   */
  bottom?: number;
  /**
   * 消息位置
   */
  placement?: MessagePlacement;
  /**
   * 包含 toast 和 notification 的容器
   */
  getContainer?: () => HTMLElement;
}

export interface MessageConfig {
  key?: React.Key;
  transitionProps?: CSSTransitionProps;
  duration?: number | null;
  placement?: MessagePlacement;
  onClose?: () => void;
  renderMessage: (args: RenderMessageArgs) => React.ReactNode;
}

interface MessageConfigState extends MessageConfig {
  key: React.Key;
}

export interface NotificationManagerState {
  destroied: boolean;
  messages: {
    topRight: MessageConfigState[];
    topLeft: MessageConfigState[];
    bottomRight: MessageConfigState[];
    bottomLeft: MessageConfigState[];
    top: MessageConfigState[];
  };
}

export function isNotAutoClose(duration?: null | number) {
  return duration === null || duration === 0;
}

function getDefaultMessages() {
  return {
    topRight: [],
    topLeft: [],
    bottomRight: [],
    bottomLeft: [],
    top: [],
  };
}

interface ManagerItemProps {
  children: React.ReactNode;
  prefixCls: string;
  duration?: number | null;
  startTimer: (duration?: number | null) => void;
  stopTimer: () => void;
}

function ManagerItem(props: ManagerItemProps) {
  const { prefixCls, duration, startTimer: startTimerProp, stopTimer, children } = props;
  const ref = useRef<HTMLDivElement>(null);
  const startTimer = useEventCallback(startTimerProp);
  useEffect(() => {
    if (ref.current && !isNotAutoClose(duration)) {
      startTimer(duration);
    }
  }, [duration, startTimer]);

  return (
    <div ref={ref} className={`${prefixCls}-item`} onMouseEnter={stopTimer} onMouseLeave={() => startTimer()}>
      {children}
    </div>
  );
}

/**
 * 包含五个方位的 managers
 */
class NotificationManager extends React.Component<NotificationManagerProps, NotificationManagerState> {
  timers: { [key: string]: number } = {};
  private containerRef = createRef<HTMLDivElement>();
  private classPrefix = '';

  constructor(props: NotificationManagerProps) {
    super(props);
    this.state = {
      destroied: false,
      messages: getDefaultMessages(),
    };
  }

  /**
   * 全局 config
   */
  static config: NotificationManagerProps = {};

  /**
   * 计算后的配置
   */
  get config() {
    return Object.assign({}, NotificationManager.config, this.props);
  }

  /**
   * 优先使用 add 方法的 config，其次是已经存在的 config，最后降级到 this.props 的配置
   * @param props
   * @param existIdx
   * @returns
   */
  getMessageConfig(props: MessageConfig, existIdx: number, placement: MessagePlacement): MessageConfigState {
    const key = props.key || getIdentity();
    const placementCfg = props.placement || placement;
    const { transitionProps, duration: durationProps } = this.config;
    const { messages } = this.state;
    const prevNotice: MessageConfigState | undefined = messages[placementCfg][existIdx];

    let duration: MessageConfigState['duration'];
    if (props.duration !== undefined) {
      duration = props.duration; // eslint-disable-line prefer-destructuring
    } else if (prevNotice?.duration !== undefined) {
      duration = prevNotice.duration; // eslint-disable-line prefer-destructuring,react/destructuring-assignment
    } else {
      duration = durationProps;
    }

    return {
      key,
      duration,
      placement,
      transitionProps: props.transitionProps || prevNotice?.transitionProps || transitionProps,
      onClose: props.onClose || prevNotice?.onClose,
      renderMessage: props.renderMessage || prevNotice?.renderMessage,
    };
  }

  updateTimers(msgs: MessageConfigState[]) {
    const { maxCount } = this.config;
    if (msgs) {
      // 隐藏的 message 关闭 timer，避免集体消失
      msgs.map((m, idx) => {
        if (!maxCount) {
          if (!this.timers[m.key]) {
            this.startTimer(m.key, m.duration);
          }
        } else if (maxCount > idx) {
          // 1. 首次插入的消息
          // 2. 首次从隐藏进入显示的消息
          if (!this.timers[m.key]) {
            this.startTimer(m.key, m.duration);
          }
        } else {
          this.stopTimer(m.key);
        }
        return null;
      });
    }
  }

  add(props: MessageConfig) {
    const { messages } = this.state;
    const { placement = DEFAULT_PLACEMENT, maxCount } = this.config;
    const placementCfg = props.placement || placement;
    let existIdx = -1;
    if (props.key) {
      existIdx = messages[placementCfg].findIndex(m => m.key === props.key);
    }
    const config = this.getMessageConfig(props, existIdx, placementCfg);
    const { key } = config;
    const newMessages = messages[placementCfg].concat();
    if (existIdx > -1) {
      this.stopTimer(config.key);
      newMessages[existIdx] = config;
    } else {
      newMessages.push(config);
      maxCount && newMessages.length > maxCount && newMessages.shift();
    }
    messages[placementCfg] = newMessages;

    this.setState({ messages }, () => this.updateTimers(newMessages));
    return key;
  }

  remove(key: React.Key): void {
    const { messages } = this.state;
    let foundMsg: MessageConfigState | null = null;
    const newMessages: NotificationManagerState['messages'] = getDefaultMessages();
    const placements = Object.keys(messages) as MessagePlacement[];
    for (let i = 0; i < placements.length; i++) {
      const placement = placements[i];
      const msgs = messages[placement];
      const msg = msgs?.find(m => m.key === key);
      if (msg) {
        foundMsg = msg;
        newMessages[placement] = msgs.filter(v => v.key !== key);
      } else {
        newMessages[placement] = msgs;
      }
    }

    if (foundMsg) {
      const { placement } = foundMsg;
      this.stopTimer(key);
      this.setState({ messages: newMessages }, () => placement && this.updateTimers(newMessages[placement]));
    }
  }

  /**
   * 清除除了 container 以外的
   */
  destroy(): void {
    Object.keys(this.timers).forEach(key => this.stopTimer(key));
    this.setState({ messages: getDefaultMessages(), destroied: true });
  }

  handleClose(message: MessageConfig) {
    message.key && this.remove(message.key);
  }

  startTimer = (key: React.Key, duration?: number | null) => {
    if (isNotAutoClose(duration)) {
      return;
    }
    const timer = this.timers[key];
    if (timer) {
      window.clearTimeout(timer);
    }
    this.timers[key] = window.setTimeout(() => this.remove(key), duration ?? DEFAULT_CLOSE_DURATION);
    return timer;
  };

  stopTimer = (key: React.Key) => {
    const timer = this.timers[key];
    if (timer) {
      window.clearTimeout(timer);
    }
    delete this.timers[key];
  };

  renderMessage(message: MessageConfigState, rootPrefixCls: string) {
    const { key, duration, renderMessage, onClose, transitionProps: messageTransitionProps } = message;
    const { transitionProps, className } = this.config;
    const transProps = {
      timeout: DEFAULT_TRANSITION_DURATION,
      ...transitionProps,
      ...messageTransitionProps,
    } as CSSTransitionProps;
    const startTimer = (newDration?: number | null) => this.startTimer(key, newDration || duration);
    const stopTimer = () => this.stopTimer(key);
    return (
      <CSSTransition
        {...transProps}
        key={getEscapedKey(key)}
        onEnter={node => {
          /**
           * 当重复调用 open/close 时，上次 exit 可能没完成
           * 这里清除一下上次残留的状态
           */
          node.style.height = '';
        }}
        onExit={(node: HTMLElement) => {
          node.style.height = `${node.scrollHeight}px`;
        }}
        onExiting={node => {
          node.style.height = '0';
        }}
        onExited={() => {
          onClose && onClose();
        }}
      >
        {status => (
          <ManagerItem prefixCls={this.classPrefix} stopTimer={stopTimer} startTimer={startTimer} duration={duration}>
            {/* /src/message/Toast.tsx getManagerConfig */}
            {renderMessage({
              close: () => this.handleClose(message),
              transitionStatus: status,
              startTimer,
              stopTimer,
              rootPrefixCls,
              toastConfigCls: className,
            })}
          </ManagerItem>
        )}
      </CSSTransition>
    );
  }

  renderManager(placement: MessagePlacement) {
    const { messages } = this.state;
    const msgs = messages[placement];

    const { className, style, top, bottom, prefixCls, getContainer } = this.config;

    const container = getContainer?.();
    const mergeStyle = {
      ...style,
      top,
      bottom,
    };

    return (
      <ManagerWrapConsumer key={placement}>
        {toastGlobalConfig => {
          const { managerGlobalToast, managerGlobalRootPrefixCls } = toastGlobalConfig;
          if (!this.classPrefix) {
            this.classPrefix = getCssScopeFormatter(managerGlobalRootPrefixCls, 'msg-manager', prefixCls)();
          }
          const cls = clsx(className, this.classPrefix, `${this.classPrefix}--${placement}`, {
            [`${this.classPrefix}--sticky`]: !container,
          });
          return (
            <div
              ref={this.containerRef}
              className={cls}
              style={`${placement}` === 'top' ? merge({ top: managerGlobalToast?.top }, mergeStyle) : mergeStyle}
              data-testid={`${this.classPrefix}-${placement}`}
            >
              <TransitionGroup>
                {msgs.map(message => this.renderMessage(message, managerGlobalRootPrefixCls))}
              </TransitionGroup>
            </div>
          );
        }}
      </ManagerWrapConsumer>
    );
  }

  render() {
    const { messages = [] } = this.state;

    const { getContainer } = this.config;
    const container = getContainer?.();

    const managers = <>{Object.keys(messages).map(placement => this.renderManager(placement as MessagePlacement))}</>;
    if (container) {
      return ReactDOM.createPortal(managers, container);
    }
    return managers;
  }
}

export default NotificationManager;
