// @universe-design/react/src/notification/notice.tsx
import type { CSSProperties, ReactNode } from 'react';
import React, { forwardRef, useRef, useEffect } from 'react';
import SucceedFilled from '@universe-design/icons-react/SucceedFilled';
import WarningFilled from '@universe-design/icons-react/WarningFilled';
import ErrorFilled from '@universe-design/icons-react/ErrorFilled';
import InfoFilled from '@universe-design/icons-react/InfoFilled';
import CloseOutlined from '@universe-design/icons-react/CloseOutlined';
import clsx from 'clsx';
import stylex from '@universe-design/stylex';
import { cssTypographyTitle4, cssTypographyBody0 } from '../stylex';
import { useCssScope } from '../config-provider/hooks';
import { Button } from '../button';
import type { PresetStatusColorType } from '../util';

const typeIcons = {
  success: SucceedFilled,
  warning: WarningFilled,
  info: InfoFilled,
  error: ErrorFilled,
};

export interface NoticeProps {
  prefixCls?: string;
  style?: CSSProperties;
  className?: string;
  closeIcon?: ReactNode;
  closable?: boolean;
  message: React.ReactNode;
  description?: React.ReactNode;
  /**
   * 自定义关闭按钮
   */
  btn?: React.ReactNode;
  /**
   * TODO:
   * 支持替换整个 content
   */
  content?: React.ReactNode;
  type?: PresetStatusColorType;
  /**
   * 替换状态按钮。
   */
  icon?: React.ReactNode;
  onClose?: React.MouseEventHandler<HTMLElement>;
  /**
   * `mouseenter` 事件。
   */
  onMouseEnter?: React.MouseEventHandler<HTMLDivElement>;
  /**
   * `mouseleave` 事件。
   */
  onMouseLeave?: React.MouseEventHandler<HTMLDivElement>;
  /**
   * `click` 事件。
   */
  onClick?: React.MouseEventHandler<HTMLDivElement>;
}

export const Notice = forwardRef<HTMLDivElement, NoticeProps>(
  (
    {
      btn,
      prefixCls,
      className,
      closable = true,
      closeIcon,
      message,
      description,
      style,
      icon,
      type = 'info',
      content,
      onClick,
      onMouseEnter,
      onMouseLeave,
      onClose,
    },
    ref,
  ) => {
    const defaultScope = useCssScope('notification');
    const cssScope = prefixCls || defaultScope();
    const callback = useRef(onClose);
    const cls = clsx(cssScope, className, {
      [`${cssScope}__notice`]: true,
      [`${cssScope}-closable`]: closable,
    });

    useEffect(() => {
      callback.current = onClose;
    }, [onClose]);

    const onRequestClose = (e: React.MouseEvent<HTMLElement>) => {
      onClose && onClose(e);
    };

    const renderIcon = React.useCallback(() => {
      if (icon) {
        return icon;
      } else if (type) {
        const Icon = typeIcons[type];
        return <Icon className={`${cssScope}__notice-content-icon`} />;
      }
    }, [icon, type, cssScope]);

    const renderClose = () => {
      if (closable) {
        return (
          <span className={`${cssScope}__notice-close`} tabIndex={0} onClick={onRequestClose}>
            {closeIcon || (
              <Button className={`${cssScope}__notice-close-icon`} icon={<CloseOutlined />} type="icon" size="small" />
            )}
          </span>
        );
      }
      return null;
    };

    return (
      <div
        className={cls}
        onClick={onClick}
        onMouseLeave={onMouseLeave}
        onMouseEnter={onMouseEnter}
        ref={ref}
        data-testid={cssScope}
      >
        {content || (
          <div style={style} className={clsx(`${cssScope}__notice-content`, `${cssScope}__notice-${type}`)}>
            <span className={`${cssScope}__notice-border`} />
            <div className={`${cssScope}__notice-icon-${type ? type : 'default'}`}>{renderIcon()}</div>
            <div className={`${cssScope}__notice-with-icon`}>
              <div className={clsx(`${cssScope}__notice-title`, stylex(cssTypographyTitle4))}>{message}</div>
              <div className={clsx(`${cssScope}__notice-des`, stylex(cssTypographyBody0))}>{description}</div>
            </div>
            {btn && <div className={`${cssScope}__notice-operation`}>{btn}</div>}
            {renderClose()}
          </div>
        )}
      </div>
    );
  },
);
