// @universe-design/react/src/notification/Notification.tsx
import type { ReactChild } from 'react';
import React, { createRef } from 'react';
import pick from 'lodash/pick';
import omit from 'lodash/omit';
import type { CSSTransitionProps } from 'react-transition-group/CSSTransition';
import type { PresetStatusColorType } from '../util';
import { Notice } from './notice';
import type { MessageConfig, MessagePlacement } from './manager';
import NotificationManager from './manager';

/**
 * 所有命令式调用方法的第一个参数
 */
export type NotificationConfigOrContentType = ReactChild | null | undefined | NotificationInstanceMethod;

const ManagerProps = [
  'className',
  'prefixCls',
  'duration',
  'transitionProps',
  'maxCount',
  'getContainer',
  'top',
  'bottom',
  'placement',
];

export interface NotificationProps {
  /**
   * 是否可关闭
   * @defaultValue false
   */
  closable?: boolean;
  /**
   * @deprecated UD 不支持关闭按钮替换
   */
  closeIcon?: React.ReactNode;
  /**
   * 是否显示图标，这里和 Other DS 不同，UD 默认为 true。UD 规范不推荐使用
   * @deprecated 已废弃，未实现
   * @internal
   */
  showIcon?: boolean;
  /**
   * `style` 属性透传。
   */
  style?: React.CSSProperties;
  /**
   *  `className` 属性透传。
   */
  className?: string;
  /**
   * 替换默认的 class prefix
   */
  prefixCls?: string;
  /**
   * 单位为毫秒，默认 4000 毫秒后自动关闭，配置为 0 或 null 则不自动关闭
   * @defaultValue 4000
   */
  duration?: number | null;
  /**
   * 自定义 CSSTransition 的 Props
   */
  transitionProps?: CSSTransitionProps;
  /**
   * 最多显示多少个消息，超过的消息将在队列中维护
   */
  maxCount?: number;
  /**
   * 包含 DOM 元素插入的容器，不提供则默认渲染在当前容器
   */
  getContainer?: () => HTMLElement;
  /**
   * 消息从顶部弹出时，距离顶部的位置，单位为像素
   * @defaultValue 24
   */
  top?: number;
  /**
   * 消息从底部弹出时，距离底部的位置，单位为像素
   * @defaultValue 24
   */
  bottom?: number;
  /**
   * 消息在窗口的位置
   * @defaultValue `topRight`
   */
  placement?: MessagePlacement;
  /**
   * @deprecated RTL 模式，暂时不支持
   */
  rtl?: boolean;
}

/**
 * @docgen
 */
export type NotificationConfigProps = Pick<
  NotificationProps,
  'top' | 'bottom' | 'duration' | 'prefixCls' | 'className' | 'placement' | 'getContainer' | 'closeIcon' | 'rtl'
>;

/**
 * @docgen
 */
export interface NotificationInstanceProps extends Omit<NotificationProps, 'top' | 'bottom'> {
  /**
   * 通知提醒标题
   */
  message: React.ReactNode;
  /**
   * 通知提醒说明
   */
  description?: React.ReactNode;
  /**
   * 替换关闭按钮
   */
  btn?: React.ReactNode;
  /**
   * 支持替换整个消息内容
   */
  content?: React.ReactNode;
  /**
   * 当前通知唯一标志，可用于更新和销毁
   */
  key?: string;
  /**
   * 通知类型
   */
  type?: PresetStatusColorType;
  /**
   * 替换状态按钮。
   */
  icon?: React.ReactNode;
  /**
   * 通知关闭的事件回调
   */
  onClose?: () => void;
  /**
   * 通知的 `onMouseEnter` 事件回调。
   */
  onMouseEnter?: React.MouseEventHandler<HTMLDivElement>;
  /**
   * 通知的 `onMouseLeave` 事件回调。
   */
  onMouseLeave?: React.MouseEventHandler<HTMLDivElement>;
  /**
   * 通知的 `onClick` 事件回调。
   */
  onClick?: React.MouseEventHandler<HTMLDivElement>;
}

export type NotificationInstanceMethod = Omit<NotificationInstanceProps, 'type'>;

export type ConfigDuration = NotificationProps['duration'] | (() => NotificationProps['duration']);

export function createNotificationWrapper(getInstance: () => Promise<Notification>) {
  return {
    /** 打开通知 */
    open(config: NotificationInstanceProps): Promise<void> {
      return getInstance().then(t => t.open(config));
    },
    /** 信息通知 */
    info(
      configOrContent: NotificationConfigOrContentType,
      duration?: ConfigDuration,
      onClose?: () => void,
    ): Promise<void> {
      return getInstance().then(t => t.info(configOrContent, duration, onClose));
    },
    /** 成功通知 */
    success(
      configOrContent: NotificationConfigOrContentType,
      duration?: ConfigDuration,
      onClose?: () => void,
    ): Promise<void> {
      return getInstance().then(t => t.success(configOrContent, duration, onClose));
    },
    /** 警告通知 */
    warning(
      configOrContent: NotificationConfigOrContentType,
      duration?: ConfigDuration,
      onClose?: () => void,
    ): Promise<void> {
      return getInstance().then(t => t.warning(configOrContent, duration, onClose));
    },
    /** 同 `warning` */
    warn(
      configOrContent: NotificationConfigOrContentType,
      duration?: ConfigDuration,
      onClose?: () => void,
    ): Promise<void> {
      return getInstance().then(t => t.warn(configOrContent, duration, onClose));
    },
    /** 错误通知 */
    error(
      configOrContent: NotificationConfigOrContentType,
      duration?: ConfigDuration,
      onClose?: () => void,
    ): Promise<void> {
      return getInstance().then(t => t.error(configOrContent, duration, onClose));
    },
    /** 通知全局配置 */
    config(options: NotificationConfigProps): void {
      getInstance().then(t => t.config(options));
    },
    /** 同 `destroy` */
    close(key?: React.Key) {
      getInstance().then(t => t.close(key));
    },
    /** 按 `key` 销毁通知，或销毁所有通知 */
    destroy(key?: React.Key) {
      getInstance().then(t => t.destroy(key));
    },
  };
}

export type NotificationInstance = ReturnType<typeof createNotificationWrapper>;

function getManagerConfig(option: NotificationInstanceProps): MessageConfig {
  const { key, duration, onClose, placement, ...rest } = option;

  return {
    key,
    // 如果自定义了操作按钮，则不自动关闭
    duration: rest.btn ? null : duration,
    placement,
    renderMessage: ({ close }) => <Notice {...rest} onClose={close} />,
    onClose,
  };
}

export default class Notification extends React.PureComponent<NotificationProps> implements NotificationInstance {
  managerRef = createRef<NotificationManager>();

  /**
   * 全局 config
   */
  static globalConfig: NotificationProps = {};

  /**
   * 计算后的配置
   */
  private get _config() {
    return Object.assign({}, Notification.globalConfig, this.props);
  }

  /* eslint-disable no-dupe-class-members */
  open(config: NotificationInstanceProps): Promise<void> {
    const manager = this.managerRef.current;
    const handleClose = config.onClose;
    return new Promise(resolve => {
      config.onClose = () => {
        handleClose && handleClose();
        resolve(undefined);
      };
      manager?.add(this._getManagerConfig(config)) || '';
    });
  }

  info(
    configOrContent: NotificationConfigOrContentType,
    duration?: ConfigDuration,
    onClose?: () => void,
  ): Promise<void> {
    return this._open('info', configOrContent, duration, onClose);
  }

  success(
    configOrContent: NotificationConfigOrContentType,
    duration?: ConfigDuration,
    onClose?: () => void,
  ): Promise<void> {
    return this._open('success', configOrContent, duration, onClose);
  }

  warning(
    configOrContent: NotificationConfigOrContentType,
    duration?: ConfigDuration,
    onClose?: () => void,
  ): Promise<void> {
    return this._open('warning', configOrContent, duration, onClose);
  }

  warn(
    configOrContent: NotificationConfigOrContentType,
    duration?: ConfigDuration,
    onClose?: () => void,
  ): Promise<void> {
    return this.warning(configOrContent, duration, onClose);
  }

  error(
    configOrContent: NotificationConfigOrContentType,
    duration?: ConfigDuration,
    onClose?: () => void,
  ): Promise<void> {
    return this._open('error', configOrContent, duration, onClose);
  }

  config(props: NotificationConfigProps): void {
    Notification.globalConfig = Object.assign(Notification.globalConfig, props);
    // Notification.config = props;
    this.forceUpdate();
  }

  destroy(key?: React.Key) {
    if (key) {
      this.managerRef.current?.remove(key);
    } else {
      this.managerRef.current?.destroy();
    }
  }

  close(key?: React.Key) {
    return this.destroy(key);
  }

  render() {
    return (
      <NotificationManager
        ref={this.managerRef}
        transitionProps={{
          timeout: 400,
          classNames: 'notification-animation',
        }}
        duration={4000}
        {...pick(this._config, ManagerProps)}
      />
    );
  }

  private _getManagerConfig(config: NotificationInstanceProps) {
    const props = Object.assign(omit(this.props, ManagerProps), omit(this._config, ManagerProps));
    return getManagerConfig({
      ...props,
      ...config,
    });
  }

  private _open(
    type: NotificationInstanceProps['type'],
    configOrContent: React.ReactNode | NotificationInstanceMethod,
    duration?: ConfigDuration,
    onClose?: () => void,
  ): Promise<void> {
    if (
      configOrContent &&
      typeof configOrContent === 'object' &&
      (('message' in configOrContent) as unknown as NotificationInstanceMethod)
    ) {
      return this.open({
        ...(configOrContent as NotificationInstanceMethod),
        type,
      });
    } else {
      const d = typeof duration === 'function' ? duration() : duration;
      return this.open({
        type,
        message: configOrContent,
        duration: d,
        onClose,
      });
    }
  }
  /* eslint-enable no-dupe-class-members */
}
