// @universe-design/react/src/notification/TransitionGroup.tsx
/**
 * This file copy most code from react-transition-group
 */
import React, { isValidElement } from 'react';
import TransitionGroupContext from 'react-transition-group/esm/TransitionGroupContext';

import {
  getChildMapping,
  getInitialChildMapping,
  getNextChildMapping,
} from 'react-transition-group/esm/utils/ChildMapping';

interface ChildrenMap {
  [key: string]: React.ReactElement;
}

function getElem(mapping: ChildrenMap, key: React.Key): React.ReactElement {
  return mapping && mapping[`.$${key}`];
}

function isElemIn(mapping: ChildrenMap, key: React.Key) {
  const elem = getElem(mapping, key);
  return elem && elem.props && elem.props.in;
}

/**
 * // 删除中间元素
 * case1: [1, 2, 3] + [1, 3] => [1, 2, 3]
 * // 在前面插入
 * case2: [1, 2, 3] + [4, 1, 2, 3] => [4, 1, 2, 3]
 * // 在后面插入
 * case3: [1, 2, 3] + [1, 2, 3, 4] => [1, 2, 3, 4]
 * // 在前面插入，删除中间元素
 * case4: [1, 2, 3] + [4, 1, 3] => [4, 1, 2, 3]
 * // 删除第一个，并新增元素
 * case5: [1, 2, 3] + [2, 3, 4] => [1, 2, 3, 4]
 */
function mergeChildrenArray(
  prevChildrenKeys: React.Key[],
  prevMapping: ChildrenMap,
  nextChildren: React.ReactNode,
): React.Key[] {
  const prevKeys = prevChildrenKeys || [];
  const next = nextChildren instanceof Array ? nextChildren : [nextChildren];
  const newChildrenSort = prevKeys.filter(k => isElemIn(prevMapping, k));
  const firstPrevChildKey = newChildrenSort[0];
  // 判断 prevKeys 中的第一个 key 是否已经遍历过。
  // 如果 nextChildren 中删除了 prevKeys 中的第一个 key，则认为已经遍历过。
  let firstPassed = Boolean(
    next.find(c => {
      if (isValidElement(c)) {
        return c.key !== firstPrevChildKey;
      }
      return false;
    }),
  );
  for (let i = 0; i < next.length; i++) {
    const child = next[i];
    if (!child || !isValidElement(child) || !child.key) {
      continue;
    }
    if (newChildrenSort.find(k => k === child.key)) {
      if (firstPrevChildKey === child.key) {
        firstPassed = true;
      }
      continue;
    }
    if (firstPassed) {
      // prevKeys 中第一个 key 遍历过，则 push 到新数组
      newChildrenSort.push(child.key);
    } else {
      newChildrenSort.unshift(child.key);
    }
  }
  return newChildrenSort;
}

interface TransitionGroupProps {
  /**
   * `<TransitionGroup>` renders a `<div>` by default. You can change this
   * behavior by providing a `component` prop.
   * If you use React v16+ and would like to avoid a wrapping `<div>` element
   * you can pass in `component={null}`. This is useful if the wrapping div
   * borks your css styles.
   */
  component?: React.ComponentType<React.PropsWithChildren<unknown>>;
  /**
   * A set of `<Transition>` components, that are toggled `in` and out as they
   * leave. the `<TransitionGroup>` will inject specific transition props, so
   * remember to spread them through if you are wrapping the `<Transition>` as
   * with our `<Fade>` example.
   *
   * While this component is meant for multiple `Transition` or `CSSTransition`
   * children, sometimes you may want to have a single transition child with
   * content that you want to be transitioned out and in when you change it
   * (e.g. routes, images etc.) In that case you can change the `key` prop of
   * the transition child as you change its content, this will cause
   * `TransitionGroup` to transition the child out and back in.
   */
  children: React.ReactNode;

  /**
   * A convenience prop that enables or disables appear animations
   * for all children. Note that specifying this will override any defaults set
   * on individual children Transitions.
   */
  appear?: boolean;
  /**
   * A convenience prop that enables or disables enter animations
   * for all children. Note that specifying this will override any defaults set
   * on individual children Transitions.
   */
  enter?: boolean;
  /**
   * A convenience prop that enables or disables exit animations
   * for all children. Note that specifying this will override any defaults set
   * on individual children Transitions.
   */
  exit?: boolean;

  /**
   * You may need to apply reactive updates to a child as it is exiting.
   * This is generally done by using `cloneElement` however in the case of an exiting
   * child the element has already been removed and not accessible to the consumer.
   *
   * If you do need to update a child as it leaves you can provide a `childFactory`
   * to wrap every child, even the ones that are leaving.
   *
   */
  childFactory?: (child: React.ReactElement) => React.ReactElement;
}

interface TransitionGroupState {
  contextValue: { isMounting: boolean };
  handleExited: (child: React.ReactElement, node: HTMLElement) => void;
  firstRender: boolean;
  childrenSort: string[];
  children: ChildrenMap;
}

const defaultProps = {
  component: 'div',
  childFactory: (child: React.ReactElement) => child,
};

/**
 * The `<TransitionGroup>` component manages a set of transition components
 * (`<Transition>` and `<CSSTransition>`) in a list. Like with the transition
 * components, `<TransitionGroup>` is a state machine for managing the mounting
 * and unmounting of components over time.
 *
 * Consider the example below. As items are removed or added to the TodoList the
 * `in` prop is toggled automatically by the `<TransitionGroup>`.
 *
 * Note that `<TransitionGroup>`  does not define any animation behavior!
 * Exactly _how_ a list item animates is up to the individual transition
 * component. This means you can mix and match animations across different list
 * items.
 */
class TransitionGroup extends React.Component<TransitionGroupProps, TransitionGroupState> {
  private mounted = false;

  constructor(props: TransitionGroupProps) {
    super(props);

    // Initial children should all be entering, dependent on appear
    this.state = {
      contextValue: { isMounting: true },
      handleExited: this.handleExited,
      firstRender: true,
      childrenSort: [],
      children: {},
    };
  }

  static getDerivedStateFromProps(
    nextProps: TransitionGroupProps,
    { children: prevChildMapping, handleExited, firstRender, childrenSort }: TransitionGroupState,
  ) {
    return {
      childrenSort: mergeChildrenArray(childrenSort, prevChildMapping, nextProps.children),
      children: firstRender
        ? getInitialChildMapping(nextProps, handleExited)
        : getNextChildMapping(nextProps, prevChildMapping, handleExited),
      firstRender: false,
    };
  }

  componentDidMount() {
    this.mounted = true;
    this.setState({
      contextValue: { isMounting: false },
    });
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  // node is `undefined` when user provided `nodeRef` prop
  handleExited = (child: React.ReactElement, node: HTMLElement) => {
    const { children } = this.props;
    const currentChildMapping = getChildMapping(children);

    if (child.key && child.key in currentChildMapping) {
      return;
    }

    if (child.props.onExited) {
      child.props.onExited(node);
    }

    if (this.mounted) {
      this.setState(state => {
        const mapping = { ...state.children };

        if (child.key) {
          delete mapping[child.key];
        }
        return { children: mapping };
      });
    }
  };

  render() {
    const {
      component: Component = defaultProps.component,
      childFactory = defaultProps.childFactory,
      ...props
    } = this.props;
    const { contextValue, childrenSort, children: mapping } = this.state;
    const children = childrenSort.map(key => {
      const elem = getElem(mapping, key);
      return childFactory(elem);
    });

    delete props.appear;
    delete props.enter;
    delete props.exit;

    if (Component === null) {
      return <TransitionGroupContext.Provider value={contextValue}>{children}</TransitionGroupContext.Provider>;
    }
    return (
      <TransitionGroupContext.Provider value={contextValue}>
        <Component {...props}>{children}</Component>
      </TransitionGroupContext.Provider>
    );
  }
}

export default TransitionGroup;
