// @universe-design/styles/esm/vc/index.light.atom.js
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
export const vctoken_meetin_bg_video_off = '#f1f3f4';
export const vctoken_meeting_bg_action_panel = 'rgba(242, 243, 245, 0.98)';
export const vctoken_meeting_bg_base = '#f4f5f6';
export const vctoken_meeting_bg_cam_off = '#dee0e3';
export const vctoken_meeting_bg_chat_bubbles_blue_focus = '#bccce5';
export const vctoken_meeting_bg_chat_bubbles_grey = 'rgba(31, 35, 41, 0.06)';
export const vctoken_meeting_bg_chat_bubbles_grey_focus = 'rgba(31, 35, 41, 0.12)';
export const vctoken_meeting_bg_cohost = 'rgba(143, 149, 158, 0.8)';
export const vctoken_meeting_bg_feed = '#1f2329';
export const vctoken_meeting_bg_float = '#ffffff';
export const vctoken_meeting_bg_float_hover = '#f5f6f7';
export const vctoken_meeting_bg_float_mic = 'rgba(100, 106, 115, 0.8)';
export const vctoken_meeting_bg_float_pressed = '#eff0f1';
export const vctoken_meeting_bg_float_transparent = 'rgba(255, 255, 255, 0.85)';
export const vctoken_meeting_bg_float_video = '#e8e9eb';
export const vctoken_meeting_bg_host = 'rgba(81, 86, 93, 0.85)';
export const vctoken_meeting_bg_host_tag = 'rgba(31, 35, 41, 0.8)';
export const vctoken_meeting_bg_mobile_share_bar = '#eff0f1';
export const vctoken_meeting_bg_mobile_share_bar_landscape = 'rgba(242, 243, 245, 0.9)';
export const vctoken_meeting_bg_ongoing = '#d0f5ce';
export const vctoken_meeting_bg_ongoing_hover = '#ffffff';
export const vctoken_meeting_bg_pc_ipad_scroll_tips = '#ebecee';
export const vctoken_meeting_bg_preview_cam_off = '#eaedf3';
export const vctoken_meeting_bg_push = 'rgba(255, 255, 255, 0.94)';
export const vctoken_meeting_bg_share_bar_immersive_mode = 'rgba(222, 224, 227, 0.9)';
export const vctoken_meeting_bg_share_panel = '#f5f6f7';
export const vctoken_meeting_bg_share_screen_pause = '#ffb679';
export const vctoken_meeting_bg_share_screen_sharing = '#85dc8e';
export const vctoken_meeting_bg_side_by_side_view = '#ebecee';
export const vctoken_meeting_bg_subtitle = 'rgba(255, 255, 255, 0.8)';
export const vctoken_meeting_bg_toolbar = 'rgba(255, 255, 255, 0.94)';
export const vctoken_meeting_bg_upcoming = '#e0e9ff';
export const vctoken_meeting_bg_upcoming_hover = '#ffffff';
export const vctoken_meeting_bg_video_call = '#2b2f36';
export const vctoken_meeting_bg_video_off_auto_layout = '#e9ebed';
export const vctoken_meeting_btn_bg_minutes_m = '#e9eafb';
export const vctoken_meeting_btn_bg_minutes_s = '#ffffff';
export const vctoken_meeting_btn_bg_new_meeting_m = '#e0e9ff';
export const vctoken_meeting_btn_bg_new_meeting_s = '#ffffff';
export const vctoken_meeting_btn_bg_on_gray = '#646a73';
export const vctoken_meeting_btn_bg_schedule_m = '#fee7cd';
export const vctoken_meeting_btn_bg_schedule_s = '#ffffff';
export const vctoken_meeting_btn_bg_share_screen_m = '#d0f5ce';
export const vctoken_meeting_btn_bg_share_screen_pause = '#3b1a02';
export const vctoken_meeting_btn_bg_share_screen_s = '#ffffff';
export const vctoken_meeting_btn_bg_share_screen_sharing = '#022c07';
export const vctoken_meeting_btn_bg_toolbar = '#f1f3f4';
export const vctoken_meeting_btn_icon_minutes = '#5b65f5';
export const vctoken_meeting_btn_icon_new_meeting = '#336df4';
export const vctoken_meeting_btn_icon_schedule = '#ff811a';
export const vctoken_meeting_btn_icon_share_screen = '#5cd168';
export const vctoken_meeting_fill_bar = 'rgba(31, 35, 41, 0.05)';
export const vctoken_meeting_fill_mask = 'rgba(0, 0, 0, 0.25)';
export const vctoken_meeting_icon_disable_on_gray_bg = 'rgba(43, 47, 54, 0.3)';
export const vctoken_meeting_icon_minutes = '#3ec3f7';
export const vctoken_meeting_line_divider = '#f2f3f5';
export const vctoken_meeting_line_host_tag = 'rgba(255, 255, 255, 0.15)';
export const vctoken_meeting_tag_bg_blue = 'rgba(46, 101, 209, 0.6)';
export const vctoken_meeting_tag_bg_cohost = 'rgba(143, 149, 158, 0.8)';
export const vctoken_meeting_tag_text_cohost = '#ffffff';
export const vctoken_meeting_text_button = '#5083fb';
export const vctoken_vc_annotate_bg_blue = '#3370ff';
export const vctoken_vc_annotate_bg_green = '#34c724';
export const vctoken_vc_annotate_bg_purple = '#7f3bf5';
export const vctoken_vc_annotate_bg_red = '#f54a45';
export const vctoken_vc_annotate_bg_yellow = '#ffc60a';
export const vctoken_vc_bg_body = '#f2f3f5';
export const vctoken_vc_bg_cam_off = '#dee0e3';
export const vctoken_vc_bg_card_ended = '#8f959e';
export const vctoken_vc_bg_card_left = '#2e374d';
export const vctoken_vc_bg_card_right = '#536080';
export const vctoken_vc_bg_float = '#eff0f1';
export const vctoken_vc_bg_float_2 = '#f5f6f7';
export const vctoken_vc_bg_float_video = '#dee0e3';
export const vctoken_vc_bg_mobile_share_bar = '#eff0f1';
export const vctoken_vc_bg_subtitle = '#ffffff';
export const vctoken_vc_bg_video_call = '#2b2f36';
export const vctoken_vc_btn_bg_float = '#ffffff';
export const vctoken_vc_btn_bg_float_hover = '#f2f3f5';
export const vctoken_vc_btn_bg_float_pressed = '#eff0f1';
export const vctoken_vc_btn_bg_livestream = '#c4f2ec';
export const vctoken_vc_btn_bg_minutes = '#caeffc';
export const vctoken_vc_btn_bg_new_meeting = '#e0e9ff';
export const vctoken_vc_btn_bg_opacity = 'rgba(255, 255, 255, 0.9)';
export const vctoken_vc_btn_bg_schedule = '#fee7cd';
export const vctoken_vc_btn_bg_share_screen = '#d0f5ce';
export const vctoken_vc_btn_fill_selected = 'rgba(51, 109, 244, 0.08)';
export const vctoken_vc_btn_icon_bg_livestream = '#ffffff';
export const vctoken_vc_btn_icon_bg_minutes = '#ffffff';
export const vctoken_vc_btn_icon_bg_new_meeting = '#ffffff';
export const vctoken_vc_btn_icon_bg_schedule = '#ffffff';
export const vctoken_vc_btn_icon_bg_share_screen = '#ffffff';
export const vctoken_vc_btn_icon_livestream = '#0f8575';
export const vctoken_vc_btn_icon_meeting = '#336df4';
export const vctoken_vc_btn_icon_minutes = '#047fb0';
export const vctoken_vc_btn_icon_new_meeting = '#336df4';
export const vctoken_vc_btn_icon_schedule = '#c25705';
export const vctoken_vc_btn_icon_selected = '#336df4';
export const vctoken_vc_btn_icon_share_screen = '#258832';
export const vctoken_vc_btn_text_selected = '#336df4';
export const vctoken_vc_fill_hover = 'rgba(31, 35, 41, 0.06)';
export const vctoken_vc_fill_ring = '#336df4';
export const vctoken_vc_list_icon_bg_ongoing = '#d0f5ce';
export const vctoken_vc_list_icon_bg_ongoing_hover = '#ffffff';
export const vctoken_vc_list_icon_bg_upcoming = '#e0e9ff';
export const vctoken_vc_list_icon_bg_upcoming_hover = '#ffffff';
export const vctoken_vc_shadow_sm = 'rgba(31, 35, 41, 0.5)';
export const vctoken_vc_shadow_xs = 'rgba(31, 35, 41, 0.05)';
export const vctoken_vc_share_app_border = '#eff0f1';
