// @universe-design/react/es/color-picker/components/ColorPickerColorBlock.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React from 'react';
import clsx from 'clsx';
import DoneOutlined from '@universe-design/icons-react/DoneOutlined';
import { useTokens } from '@universe-design/styles';
import { bg_float as bg_float_lm } from '@universe-design/styles/esm/all.light.atom';
import { bg_float as bg_float_dm } from '@universe-design/styles/esm/all.dark.atom';
import { ColorPickerChangeSource } from '../type';
import { formatColor } from '../util';
import { useTheme } from '../../config-provider/hooks';
import { useColorPickerContext } from '../ColorPickerContext';
import { isFunction, isNonNullable } from '../../util';

/**
 * @public
 * `<ColorPickerColorBlock />` 组件的属性。
 */

export const ColorPickerColorBlock = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    color,
    className,
    onClick,
    ...restProps
  } = _ref;
  /**
   * Context
   */
  const {
    value,
    setValue,
    showDarkCheckmark,
    rootCssScope,
    transformDisplayColor
  } = useColorPickerContext();
  const prefix = rootCssScope('-swatches');

  /**
   * Theme
   */
  const theme = useTheme();
  const {
    'bg-float': panelBgColor = theme === 'light' ? bg_float_lm : bg_float_dm
  } = useTokens();
  const colorObj = React.useMemo(() => isNonNullable(color) ? formatColor(color) : color, [color]);
  const displayColor = React.useMemo(() => isNonNullable(colorObj) ? transformDisplayColor(colorObj) : undefined, [colorObj, transformDisplayColor]);

  /**
   * Handler
   */
  const handleClick = React.useCallback(ev => {
    isFunction(onClick) && onClick(ev);
    if (isNonNullable(colorObj === null || colorObj === void 0 ? void 0 : colorObj.rgb)) {
      setValue(colorObj === null || colorObj === void 0 ? void 0 : colorObj.rgb, {
        source: ColorPickerChangeSource.COLOR_BLOCK
      });
      return;
    }
    setValue(undefined, {
      source: ColorPickerChangeSource.COLOR_BLOCK
    });
  }, [colorObj === null || colorObj === void 0 ? void 0 : colorObj.rgb, onClick, setValue]);

  /**
   * Computed state
   */
  const isActive = value && (colorObj === null || colorObj === void 0 ? void 0 : colorObj.color.equals(value));
  return /*#__PURE__*/React.createElement("div", _extends({}, restProps, {
    ref: ref,
    className: clsx(`${prefix}-cell-wrapper`, isActive && `${prefix}-cell-active`, className),
    onClick: handleClick
  }), displayColor ? /*#__PURE__*/React.createElement("div", {
    className: clsx(`${prefix}-row-cell`, displayColor.color.equals(panelBgColor) && `${prefix}-row-cell-highlight`, showDarkCheckmark(displayColor.hex, displayColor) && `${prefix}-row-cell-icon-deepen`),
    style: {
      backgroundColor: displayColor.rgbString
    }
  }, isActive ? /*#__PURE__*/React.createElement(DoneOutlined, null) : null) : /*#__PURE__*/React.createElement("div", {
    className: clsx(`${prefix}-row-cell`, `${prefix}-row-cell-highlight`, `${prefix}-row-cell-empty`)
  }, /*#__PURE__*/React.createElement("svg", {
    width: "100%",
    height: "100%",
    viewBox: "0 0 16 16"
  }, /*#__PURE__*/React.createElement("path", {
    d: "m 0 0l 16 16",
    stroke: "currentColor"
  }))));
});
