// @universe-design/react/es/color-picker/components/ColorInput.js
import React, { useEffect, useMemo, useState } from 'react';
import clsx from 'clsx';
import clamp from 'lodash/clamp';
import { useCssScope } from '../../config-provider/hooks';
import { HrgbKey } from '../type';
import NativeInput from '../../shared/components/Input';
export const ColorInput = props => {
  const {
    colorKey,
    color,
    onChange,
    prefixCls
  } = props;
  const [value, setValue] = useState(color);
  const cssScope = useCssScope('color-picker-input');
  const prefix = prefixCls || cssScope();
  useEffect(() => {
    setValue(color);
  }, [color]);
  const colorVal = useMemo(() => `${value}`.toUpperCase(), [value]);
  const judgeVal = key => {
    const val = value.toString();
    if (key !== HrgbKey.HEX) {
      if (/^-?[0-9]+\.?[0-9]*$/.test(val)) {
        // eslint-disable-next-line @typescript-eslint/no-magic-numbers
        return clamp(parseInt(val, 10), 0, 255);
      }
      return color;
    } else {
      const matchGroup = /^#?([0-9A-Fa-f]{0,6})$/.exec(val);
      if (!matchGroup) {
        return color.toString().toUpperCase();
      }
      const colorStr = matchGroup[1].toUpperCase();
      const len = colorStr.length;
      if (len !== 3) {
        return colorStr + '0'.repeat(6 - len);
      } else {
        let newVal = '';
        for (let i = 0; i < len; i++) {
          newVal += colorStr[i].repeat(2);
        }
        return newVal;
      }
    }
  };
  const handleBlur = key => {
    try {
      const newVal = judgeVal(key);
      setValue(newVal);
      onChange && onChange(key, newVal);
    } catch (err) {
      /* empty */
    }
  };
  const handeKeyUp = (e, key) => {
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    if (e.keyCode === 13) {
      handleBlur(key);
    }
  };
  return /*#__PURE__*/React.createElement("div", {
    className: clsx(prefix, colorKey === HrgbKey.HEX && `${prefix}-hex`)
  }, /*#__PURE__*/React.createElement("span", {
    className: `${prefix}-title`
  }, colorKey.toUpperCase()), /*#__PURE__*/React.createElement("div", {
    className: `${prefix}-outer`
  }, /*#__PURE__*/React.createElement(NativeInput, {
    type: "text",
    className: `${prefix}-inner`,
    value: colorVal,
    onChange: e => {
      setValue(e.target.value);
    },
    onBlur: () => handleBlur(colorKey),
    onKeyUp: e => handeKeyUp(e, colorKey)
  })));
};
