// @universe-design/react/es/color-picker/components/ColorPickerHue.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.u1euv43f{position:absolute;}
import u1u1p7mj from "./../../.stylex/u1u1p7mj";
//.ucv7f5c{top:-4px;}
import u5shkri from "./../../.stylex/u5shkri";
/** @jsxRuntime classic */
/** @jsx jsx */
import React from 'react';
import stylex from '@universe-design/stylex';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx } from '@universe-design/stylex/react';
import clsx from 'clsx';
import clamp from 'lodash/clamp';
import { useColorPickerContextSelector } from '../ColorPickerContext';
import { formatColor } from '../util';
import { ColorPickerChangeSource } from '../type';
export const ColorPickerHue = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    className,
    onHueDragEnd,
    ...props
  } = _ref;
  const setValue = useColorPickerContextSelector(o => o.setValue);
  const rootCssScope = useColorPickerContextSelector(o => o.rootCssScope);
  const transformDisplayColor = useColorPickerContextSelector(o => o.transformDisplayColor);
  const displayColor = useColorPickerContextSelector(o => o.displayColor);
  const hsv = useColorPickerContextSelector(o => o.hsv);
  const prefix = rootCssScope('-hue');
  const sliderHandlerRef = React.useRef(null);
  const calculateChange = event => {
    const el = sliderHandlerRef.current;
    if (!el) {
      return 0;
    }
    const containerWidth = el.clientWidth;
    const distance = event.pageX - (el.getBoundingClientRect().left + window.pageXOffset);
    const val = clamp(distance / containerWidth, 0, 1);
    return Math.round(val * 100);
  };
  const handleChange = event => {
    const value = calculateChange(event);
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    const h = clamp(value * 360 / 100, 0, 360);
    setValue({
      ...hsv,
      h
    }, {
      source: ColorPickerChangeSource.HUE
    });
  };
  const handlePointerUp = e => {
    onHueDragEnd === null || onHueDragEnd === void 0 ? void 0 : onHueDragEnd(e);
    window.removeEventListener('pointermove', handleChange);
    window.removeEventListener('pointerup', handlePointerUp);
  };

  // eslint-disable-next-line @typescript-eslint/no-magic-numbers
  const left = `${hsv.h * 100 / 360}%`;
  const sliderBackground = React.useMemo(() => `linear-gradient(to right, ${transformDisplayColor(formatColor('#f00')).rgbString} 0%, ${transformDisplayColor(formatColor('#ff0')).rgbString} 17%, ${transformDisplayColor(formatColor('#0f0')).rgbString} 33%, ${transformDisplayColor(formatColor('#0ff')).rgbString} 50%, ${transformDisplayColor(formatColor('#00f')).rgbString} 67%, ${transformDisplayColor(formatColor('#f0f')).rgbString} 83%, ${transformDisplayColor(formatColor('#f00')).rgbString} 100%)`, [transformDisplayColor]);
  return jsx("div", _extends({}, props, {
    ref: ref,
    className: clsx(prefix, className)
  }), jsx("div", {
    className: clsx(`${prefix}-display`, `${prefix}-display-highlight`),
    style: {
      background: displayColor && displayColor.rgbString
    }
  }), jsx("div", {
    className: `${prefix}-slider`,
    ref: sliderHandlerRef,
    onPointerDown: event => {
      handleChange(event.nativeEvent);
      window.addEventListener('pointermove', handleChange);
      window.addEventListener('pointerup', handlePointerUp);
    },
    style: {
      background: sliderBackground
    }
  }, jsx("div", {
    className: `${prefix}-pointer`,
    style: {
      left
    }
  }, jsx("svg", {
    width: 12,
    height: 5,
    sx: [{
      "props": {
        "qhf8xq": u1u1p7mj,
        "Bhzewxz": u5shkri
      }
    }]
  }, jsx("path", {
    d: "M0,4 V5 H10 V4 L5,0 Z",
    style: {
      fill: 'currentcolor'
    }
  })))));
});
