// @universe-design/react/es/color-picker/components/ColorPickerSaturation.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React, { useRef } from 'react';
import clamp from 'lodash/clamp';
import clsx from 'clsx';
import { useColorPickerContextSelector } from '../ColorPickerContext';
import composeRefs from '../../util/composeRefs';
import { isFunction } from '../../util';
import { ColorPickerChangeSource } from '../type';
import { formatColor } from '../util';

/**
 * @public
 */

/**
 * 调整饱和度组件
 */
export const ColorPickerSaturation = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    className,
    style,
    onMouseDown,
    onPointerDown,
    onSaturationDragEnd,
    ...props
  } = _ref;
  const setValue = useColorPickerContextSelector(o => o.setValue);
  const rootCssScope = useColorPickerContextSelector(o => o.rootCssScope);
  const displayColor = useColorPickerContextSelector(o => o.displayColor);
  const transformDisplayColor = useColorPickerContextSelector(o => o.transformDisplayColor);
  const hsv = useColorPickerContextSelector(o => o.hsv);
  const {
    hsv: displayHSV
  } = transformDisplayColor(formatColor(hsv));
  const prefix = rootCssScope('-saturation');
  const divRef = useRef(null);
  const mergedRef = React.useMemo(() => composeRefs(divRef, ref), [ref]);
  const position = {
    top: 100 - hsv.v * 100,
    left: hsv.s * 100
  };
  const getPosition = event => {
    const el = divRef.current;
    if (!el) {
      return position;
    }
    const containerRect = el.getBoundingClientRect();
    const {
      width: containerWidth,
      height: containerHeight
    } = containerRect;
    let left = event.pageX - (containerRect.left + window.pageXOffset);
    let top = event.pageY - (containerRect.top + window.pageYOffset);
    left = clamp(left, 0, containerWidth);
    top = clamp(top, 0, containerHeight);
    return {
      left: left * 100 / containerWidth,
      top: top * 100 / containerHeight
    };
  };
  const handleChange = event => {
    const pos = getPosition(event);
    setValue({
      h: hsv.h,
      s: pos.left / 100,
      v: 1 - pos.top / 100,
      a: hsv.a
    }, {
      source: ColorPickerChangeSource.SATURATION
    });
  };
  const handlePointerUp = e => {
    onSaturationDragEnd === null || onSaturationDragEnd === void 0 ? void 0 : onSaturationDragEnd(e);
    window.removeEventListener('pointermove', handleChange);
    window.removeEventListener('pointerup', handlePointerUp);
  };
  const bindEventListeners = () => {
    window.addEventListener('pointermove', handleChange);
    window.addEventListener('pointerup', handlePointerUp);
  };
  return /*#__PURE__*/React.createElement("div", _extends({}, props, {
    className: clsx(prefix, className),
    style: {
      ...style,
      background: displayColor && `hsl(${displayHSV.h}, 100%, 50%)`
    },
    ref: mergedRef,
    onPointerDown: event => {
      isFunction(onMouseDown) && onMouseDown(event);
      isFunction(onPointerDown) && onPointerDown(event);
      handleChange(event.nativeEvent);
      bindEventListeners();
    }
  }), /*#__PURE__*/React.createElement("div", {
    className: `${prefix}-border`
  }), /*#__PURE__*/React.createElement("div", {
    className: `${prefix}-white`
  }, /*#__PURE__*/React.createElement("div", {
    className: `${prefix}-black`
  }), /*#__PURE__*/React.createElement("div", {
    className: clsx(`${prefix}-circle`
    // 暂时先去掉这个 circle 的 judge，全部显示白色就行，这个和 https://www.google.com.hk/search?q=color+picker 的行为一致
    // judgeNeedDeepenIcon(hex) && `${prefix}-circle-deepen`
    ),
    style: {
      top: `${position.top}%`,
      left: `${position.left}%`,
      background: displayColor && displayColor.rgbString
    }
  })));
});
