// @universe-design/react/es/color-picker/components/Panel.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React from 'react';
import clsx from 'clsx';
import { ColorInput } from './ColorInput';
import { ColorPickerChangeSource, HrgbKey } from '../type';
import { COLORKEY, RGBKEY } from '../const';
import { useColorPickerContextSelector } from '../ColorPickerContext';
export const Panel = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    className,
    ...props
  } = _ref;
  const setValue = useColorPickerContextSelector(o => o.setValue);
  const rootCssScope = useColorPickerContextSelector(o => o.rootCssScope);
  const rgb = useColorPickerContextSelector(o => o.rgb);
  const hex = useColorPickerContextSelector(o => o.hex);
  const prefix = rootCssScope('-panel');
  const handleChange = (key, val) => {
    if (RGBKEY.includes(key)) {
      const newRgb = Object.assign({}, rgb, {
        [key]: val
      });
      setValue(newRgb, {
        source: ColorPickerChangeSource.INPUT
      });
      return;
    }
    setValue(val.toString(), {
      source: ColorPickerChangeSource.INPUT
    });
  };
  return /*#__PURE__*/React.createElement("div", _extends({}, props, {
    className: clsx(prefix, className),
    ref: ref
  }), (() => COLORKEY.map(key => {
    const curKeyColor = key === HrgbKey.HEX ? hex : rgb[key];
    return /*#__PURE__*/React.createElement(ColorInput, {
      colorKey: key,
      key: key,
      color: curKeyColor,
      onChange: handleChange
    });
  }))());
});
