// @universe-design/styles/esm/calendar/index.dark.atom.js
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
export const caltoken_addroms_bg_icon_1 = '#173b12';
export const caltoken_addroms_bg_icon_2 = '#21511a';
export const caltoken_addroms_bg_icon_3 = '#51ba43';
export const caltoken_btn_selected_line_gray = '#a6a6a6';
export const caltoken_event_border_unanswered_blue = '#5a7aba';
export const caltoken_event_border_unanswered_carmine = '#c24a8e';
export const caltoken_event_border_unanswered_green = '#2f7526';
export const caltoken_event_border_unanswered_indigo = '#5e68e8';
export const caltoken_event_border_unanswered_neutral = '#757575';
export const caltoken_event_border_unanswered_orange = '#b85e1a';
export const caltoken_event_border_unanswered_purple = '#8c55ec';
export const caltoken_event_border_unanswered_red = '#d14642';
export const caltoken_event_border_unanswered_tur = '#227369';
export const caltoken_event_border_unanswered_violet = '#a43da4';
export const caltoken_event_border_unanswered_wathet = '#1099cc';
export const caltoken_event_border_unanswered_yellow = '#ad7d15';
export const caltoken_event_conflict_event_shadow = '8px 8px 24px 8px rgba(0, 0, 0, 0.11), -6px 6px 12px rgba(0, 0, 0, 0.05), -4px 4px 8px -8px rgba(0, 0, 0, 0.08)';
export const caltoken_event_dark_bg_bar_blue = '#4c88ff';
export const caltoken_event_dark_bg_bar_carmine = '#db5ea4';
export const caltoken_event_dark_bg_bar_green = '#35872a';
export const caltoken_event_dark_bg_bar_indigo = '#7b83f7';
export const caltoken_event_dark_bg_bar_neutral = '#757575';
export const caltoken_event_dark_bg_bar_orange = '#db7018';
export const caltoken_event_dark_bg_bar_purple = '#a575fa';
export const caltoken_event_dark_bg_bar_red = '#f05b56';
export const caltoken_event_dark_bg_bar_tur = '#1fa18f';
export const caltoken_event_dark_bg_bar_violet = '#d661d6';
export const caltoken_event_dark_bg_bar_wathet = '#1099cc';
export const caltoken_event_dark_bg_bar_yellow = '#d49b0b';
export const caltoken_event_dark_bg_blue = '#3370eb';
export const caltoken_event_dark_bg_carmine = '#c24a8e';
export const caltoken_event_dark_bg_green = '#2f7526';
export const caltoken_event_dark_bg_indigo = '#5e68e8';
export const caltoken_event_dark_bg_neutral = '#5f5f5f';
export const caltoken_event_dark_bg_orange = '#b85e1a';
export const caltoken_event_dark_bg_pending_blue = '#275fce';
export const caltoken_event_dark_bg_pending_carmine = '#ab417d';
export const caltoken_event_dark_bg_pending_green = '#296621';
export const caltoken_event_dark_bg_pending_indigo = '#515ad6';
export const caltoken_event_dark_bg_pending_neutral = '#505050';
export const caltoken_event_dark_bg_pending_orange = '#a15317';
export const caltoken_event_dark_bg_pending_purple = '#7c4ad4';
export const caltoken_event_dark_bg_pending_red = '#b33a37';
export const caltoken_event_dark_bg_pending_tur = '#227369';
export const caltoken_event_dark_bg_pending_violet = '#a43da4';
export const caltoken_event_dark_bg_pending_wathet = '#146c91';
export const caltoken_event_dark_bg_pending_yellow = '#8a6419';
export const caltoken_event_dark_bg_purple = '#8c55ec';
export const caltoken_event_dark_bg_red = '#d14642';
export const caltoken_event_dark_bg_tur = '#198578';
export const caltoken_event_dark_bg_violet = '#b54ab5';
export const caltoken_event_dark_bg_wathet = '#1a7fab';
export const caltoken_event_dark_bg_yellow = '#ad7d15';
export const caltoken_event_dark_calendar_carmine = '#f01d94';
export const caltoken_event_dark_calendar_green = '#2ea121';
export const caltoken_event_dark_calendar_indigo = '#4954e6';
export const caltoken_event_dark_calendar_neutral = '#646a73';
export const caltoken_event_dark_calendar_orange = '#de7802';
export const caltoken_event_dark_calendar_purple = '#7f3bf5';
export const caltoken_event_dark_calendar_red = '#f54a45';
export const caltoken_event_dark_calendar_tur = '#04b49c';
export const caltoken_event_dark_calendar_violet = '#b320b3';
export const caltoken_event_dark_calendar_wathet = '#049fd7';
export const caltoken_event_dark_calendar_yellow = '#dc9b04';
export const caltoken_event_dark_pending_bar_blue = '#4c88ff';
export const caltoken_event_dark_pending_bar_carmine = '#db5ea4';
export const caltoken_event_dark_pending_bar_green = '#419e34';
export const caltoken_event_dark_pending_bar_indigo = '#7b83f7';
export const caltoken_event_dark_pending_bar_neutral = '#757575';
export const caltoken_event_dark_pending_bar_orange = '#db7018';
export const caltoken_event_dark_pending_bar_purple = '#a575fa';
export const caltoken_event_dark_pending_bar_red = '#f05b56';
export const caltoken_event_dark_pending_bar_tur = '#1fa18f';
export const caltoken_event_dark_pending_bar_violet = '#d661d6';
export const caltoken_event_dark_pending_bar_wathet = '#1099cc';
export const caltoken_event_dark_pending_bar_yellow = '#d49b0b';
export const caltoken_event_detail_bg_left_green = '#419e34';
export const caltoken_event_detail_bg_left_orange = '#db7018';
export const caltoken_event_detail_bg_left_red = '#f05b56';
export const caltoken_event_detail_bg_left_tur = '#1fa18f';
export const caltoken_event_detail_bg_left_wathet = '#1a7fab';
export const caltoken_event_detail_bg_right_green = '#51ba43';
export const caltoken_event_detail_bg_right_orange = '#f3871b';
export const caltoken_event_detail_bg_right_red = '#f05b56';
export const caltoken_event_detail_bg_right_tur = '#1ab7a1';
export const caltoken_event_detail_bg_right_wathet = '#1099cc';
export const caltoken_event_light_bg_bar_blue = '#4c88ff';
export const caltoken_event_light_bg_bar_carmine = '#db5ea4';
export const caltoken_event_light_bg_bar_green = '#419e34';
export const caltoken_event_light_bg_bar_indigo = '#7b83f7';
export const caltoken_event_light_bg_bar_neutral = '#757575';
export const caltoken_event_light_bg_bar_orange = '#b85e1a';
export const caltoken_event_light_bg_bar_purple = '#a575fa';
export const caltoken_event_light_bg_bar_red = '#d14642';
export const caltoken_event_light_bg_bar_tur = '#1fa18f';
export const caltoken_event_light_bg_bar_violet = '#d661d6';
export const caltoken_event_light_bg_bar_wathet = '#1099cc';
export const caltoken_event_light_bg_bar_yellow = '#d49b0b';
export const caltoken_event_light_bg_blue = '#173166';
export const caltoken_event_light_bg_carmine = '#591c3f';
export const caltoken_event_light_bg_green = '#173b12';
export const caltoken_event_light_bg_indigo = '#2a2d69';
export const caltoken_event_light_bg_neutral = '#373737';
export const caltoken_event_light_bg_oran = '#4a2b10';
export const caltoken_event_light_bg_pending_blue = '#152340';
export const caltoken_event_light_bg_pending_carmine = '#3a182b';
export const caltoken_event_light_bg_pending_green = '#0e2b0a';
export const caltoken_event_light_bg_pending_indigo = '#1e204a';
export const caltoken_event_light_bg_pending_neutral = '#292929';
export const caltoken_event_light_bg_pending_orange = '#33210b';
export const caltoken_event_light_bg_pending_purple = '#2b194a';
export const caltoken_event_light_bg_pending_red = '#3d1a19';
export const caltoken_event_light_bg_pending_tur = '#132926';
export const caltoken_event_light_bg_pending_violet = '#3b153b';
export const caltoken_event_light_bg_pending_wathet = '#152830';
export const caltoken_event_light_bg_pending_yellow = '#30250a';
export const caltoken_event_light_bg_purple = '#3f2073';
export const caltoken_event_light_bg_red = '#591f1d';
export const caltoken_event_light_bg_tur = '#173b36';
export const caltoken_event_light_bg_violet = '#541854';
export const caltoken_event_light_bg_wathet = '#103647';
export const caltoken_event_light_bg_yellow = '#473409';
export const caltoken_event_light_calendar_blue = '#4e83fd';
export const caltoken_event_light_calendar_carmine = '#f14ba9';
export const caltoken_event_light_calendar_green = '#34c724';
export const caltoken_event_light_calendar_indigo = '#616ae5';
export const caltoken_event_light_calendar_neutral = '#9ca2a9';
export const caltoken_event_light_calendar_orange = '#ff8800';
export const caltoken_event_light_calendar_purple = '#935af6';
export const caltoken_event_light_calendar_red = '#f76964';
export const caltoken_event_light_calendar_tur = '#00d6b9';
export const caltoken_event_light_calendar_violet = '#d136d1';
export const caltoken_event_light_calendar_wathet = '#14c0ff';
export const caltoken_event_light_calendar_yellow = '#ffc60a';
export const caltoken_event_light_text_blue = '#75a4ff';
export const caltoken_event_light_text_carmine = '#ed77ba';
export const caltoken_event_light_text_green = '#51ba43';
export const caltoken_event_light_text_indigo = '#9499f7';
export const caltoken_event_light_text_neutral = '#cfcfcf';
export const caltoken_event_light_text_orange = '#f89e44';
export const caltoken_event_light_text_purple = '#b88ffe';
export const caltoken_event_light_text_red = '#f6827e';
export const caltoken_event_light_text_tur = '#1ab7a1';
export const caltoken_event_light_text_unanswered_blue = '#6c96e8';
export const caltoken_event_light_text_unanswered_carmine = '#db5ea4';
export const caltoken_event_light_text_unanswered_green = '#419e34';
export const caltoken_event_light_text_unanswered_indigo = '#7b83f7';
export const caltoken_event_light_text_unanswered_neutral = '#a6a6a6';
export const caltoken_event_light_text_unanswered_orange = '#e08e41';
export const caltoken_event_light_text_unanswered_purple = '#a575fa';
export const caltoken_event_light_text_unanswered_red = '#dc7572';
export const caltoken_event_light_text_unanswered_tur = '#1fa18f';
export const caltoken_event_light_text_unanswered_violet = '#cd75cd';
export const caltoken_event_light_text_unanswered_wathet = '#1099cc';
export const caltoken_event_light_text_unanswered_yellow = '#d49b0b';
export const caltoken_event_light_text_violet = '#e17fe1';
export const caltoken_event_light_text_wathet = '#25b2e5';
export const caltoken_event_light_text_yellow = '#f0b622';
export const caltoken_event_quick_creation_shadow = '0px 10px 36px 10px rgba(0, 0, 0, 0.09), 0px 8px 24px rgba(0, 0, 0, 0.09), 0px 6px 12px -10px rgba(0, 0, 0, 0.11)';
export const caltoken_event_stripe_blue = '#152340';
export const caltoken_event_stripe_carmine = '#3a182b';
export const caltoken_event_stripe_green = '#0e2b0a';
export const caltoken_event_stripe_indigo = '#1e204a';
export const caltoken_event_stripe_neutral = '#292929';
export const caltoken_event_stripe_orange = '#33210b';
export const caltoken_event_stripe_purple = '#2b194a';
export const caltoken_event_stripe_red = '#3d1a19';
export const caltoken_event_stripe_tur = '#132926';
export const caltoken_event_stripe_violet = '#3b153b';
export const caltoken_event_stripe_wathet = '#152830';
export const caltoken_event_stripe_yellow = '#30250a';
export const caltoken_tag_colour_day = '#cf9500';
