// @universe-design/react/lib/date-picker/panels/PanelHeaderSwitchers.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.usePickerHeaderSwitcherProps = exports.PreYearSwitcher = exports.PreMonthSwitcher = exports.Pre20YearsSwitcher = exports.NextYearSwitcher = exports.NextMonthSwitcher = exports.Next20YearsSwitcher = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _RightBoldOutlined = _interopRequireDefault(require("@universe-design/icons-react/RightBoldOutlined"));
var _DownRightBoldOutlined = _interopRequireDefault(require("@universe-design/icons-react/DownRightBoldOutlined"));
var _LeftBoldOutlined = _interopRequireDefault(require("@universe-design/icons-react/LeftBoldOutlined"));
var _UpLeftBoldOutlined = _interopRequireDefault(require("@universe-design/icons-react/UpLeftBoldOutlined"));
var _button = require("../../button");
var _configProvider = require("../../config-provider");
const usePickerHeaderSwitcherProps = _ref => {
  let {
    className,
    asPlaceholder = false,
    ...restProps
  } = _ref;
  const prefixCls = (0, _configProvider.useCssScope)('picker')();
  const headerIconCls = `${prefixCls}-panel-header-icon`;
  return {
    ...restProps,
    type: 'icon',
    size: 'extra-small',
    className: (0, _clsx.default)(headerIconCls, asPlaceholder && `${headerIconCls}__hidden`, className)
  };
};
exports.usePickerHeaderSwitcherProps = usePickerHeaderSwitcherProps;
const NextMonthSwitcher = exports.NextMonthSwitcher = /*#__PURE__*/_react.default.forwardRef((props, ref) => {
  const buttonProps = usePickerHeaderSwitcherProps(props);
  return /*#__PURE__*/_react.default.createElement(_button.Button, (0, _extends2.default)({
    ref: ref
  }, buttonProps, {
    icon: /*#__PURE__*/_react.default.createElement(_RightBoldOutlined.default, null)
  }));
});
const NextYearSwitcher = exports.NextYearSwitcher = /*#__PURE__*/_react.default.forwardRef((props, ref) => {
  const buttonProps = usePickerHeaderSwitcherProps(props);
  return /*#__PURE__*/_react.default.createElement(_button.Button, (0, _extends2.default)({
    ref: ref
  }, buttonProps, {
    icon: /*#__PURE__*/_react.default.createElement(_DownRightBoldOutlined.default, null)
  }));
});
const Next20YearsSwitcher = exports.Next20YearsSwitcher = /*#__PURE__*/_react.default.forwardRef((props, ref) => {
  const buttonProps = usePickerHeaderSwitcherProps(props);
  return /*#__PURE__*/_react.default.createElement(_button.Button, (0, _extends2.default)({
    ref: ref
  }, buttonProps, {
    icon: /*#__PURE__*/_react.default.createElement(_DownRightBoldOutlined.default, null)
  }));
});
const PreMonthSwitcher = exports.PreMonthSwitcher = /*#__PURE__*/_react.default.forwardRef((props, ref) => {
  const buttonProps = usePickerHeaderSwitcherProps(props);
  return /*#__PURE__*/_react.default.createElement(_button.Button, (0, _extends2.default)({
    ref: ref
  }, buttonProps, {
    icon: /*#__PURE__*/_react.default.createElement(_LeftBoldOutlined.default, null)
  }));
});
const PreYearSwitcher = exports.PreYearSwitcher = /*#__PURE__*/_react.default.forwardRef((props, ref) => {
  const buttonProps = usePickerHeaderSwitcherProps(props);
  return /*#__PURE__*/_react.default.createElement(_button.Button, (0, _extends2.default)({
    ref: ref
  }, buttonProps, {
    icon: /*#__PURE__*/_react.default.createElement(_UpLeftBoldOutlined.default, null)
  }));
});
const Pre20YearsSwitcher = exports.Pre20YearsSwitcher = /*#__PURE__*/_react.default.forwardRef((props, ref) => {
  const buttonProps = usePickerHeaderSwitcherProps(props);
  return /*#__PURE__*/_react.default.createElement(_button.Button, (0, _extends2.default)({
    ref: ref
  }, buttonProps, {
    icon: /*#__PURE__*/_react.default.createElement(_UpLeftBoldOutlined.default, null)
  }));
});
