// @universe-design/react/lib/popover/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Popover = void 0;
var _u1gmpn6f = _interopRequireDefault(require("./../.stylex/u1gmpn6f"));
var _u1o367ev = _interopRequireDefault(require("./../.stylex/u1o367ev"));
var _u5p8ukz = _interopRequireDefault(require("./../.stylex/u5p8ukz"));
var _u1mtnbey = _interopRequireDefault(require("./../.stylex/u1mtnbey"));
var _u441rjq = _interopRequireDefault(require("./../.stylex/u441rjq"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _stylex2 = require("../stylex");
var _hooks = require("../config-provider/hooks");
var _InfoFilled = _interopRequireDefault(require("@universe-design/icons-react/InfoFilled"));
var _SucceedFilled = _interopRequireDefault(require("@universe-design/icons-react/SucceedFilled"));
var _WarningFilled = _interopRequireDefault(require("@universe-design/icons-react/WarningFilled"));
var _ErrorFilled = _interopRequireDefault(require("@universe-design/icons-react/ErrorFilled"));
var _Tooltip = require("../tooltip/Tooltip");
var _getRenderPropValue = require("../util/getRenderPropValue");
var _FirstLineAlign = require("../shared/components/FirstLineAlign");
require("./style");
//.u1guyajx{align-items:baseline;}
//.u4ybsrx{font-size:16px;}
//.uez10in{line-height:0;}
//.u7nlbp4{align-self:center;}
//.u19lj068{padding-right:8px;}
// 正常、宽松、无任何内容（并且无填充）
/**
 * @public
 * @docgen
 */
const iconMap = {
  success: _SucceedFilled.default,
  warning: _WarningFilled.default,
  info: _InfoFilled.default,
  error: _ErrorFilled.default
};
const Popover = exports.Popover = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  const {
    mode = 'normal',
    prefixCls,
    title,
    content,
    overlayClassName,
    icon,
    mouseLeaveDelay = 100,
    ...restProps
  } = props;
  const cssScope = (0, _hooks.useCssScope)('popover', prefixCls);
  const contentCls = cssScope('__overlay-content');
  const titleBeforeProps = (0, _react.useMemo)(() => {
    if (!icon) {
      return;
    }
    const cssIconWrapper = {
      "props": {
        "z189sj": _u441rjq.default
      }
    };
    const cssIcon = {
      "props": {
        "Be2twd7": _u1o367ev.default,
        "Bg96gwp": _u5p8ukz.default,
        "qb2dma": _u1mtnbey.default
      }
    };
    if (typeof icon === 'string' && (icon === 'success' || icon === 'warning' || icon === 'info' || icon === 'error')) {
      const Icon = iconMap[icon];
      return {
        className: `${cssScope()}-icon-${icon}`,
        sx: [cssIconWrapper],
        children: /*#__PURE__*/_react.default.createElement(Icon, {
          className: (0, _stylex.default)(cssIcon)
        })
      };
    }
    return {
      className: `${cssScope()}-icon-customer`,
      sx: [cssIconWrapper],
      children: /*#__PURE__*/_react.default.createElement("span", {
        className: (0, _clsx.default)(`${cssScope()}-icon-customer-wrapper`, (0, _stylex.default)(cssIcon))
      }, icon)
    };
  }, [cssScope, icon]);
  const popupNode = (0, _react.useMemo)(() => {
    if (!title && !content) {
      return null;
    }

    // 纯净模式只渲染 content，并且移除了 padding 相关的 style
    if (mode === 'pure') {
      return content && /*#__PURE__*/_react.default.createElement("div", {
        className: contentCls
      }, (0, _getRenderPropValue.getRenderPropValue)(content));
    }
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, title && /*#__PURE__*/_react.default.createElement(_FirstLineAlign.FirstLineAlign, {
      className: cssScope('__overlay-title'),
      sx: [_stylex2.cssFlexBase, _stylex2.cssTypographyHeadline, {
        "props": {
          "Bt984gj": _u1gmpn6f.default
        }
      }],
      beforeProps: titleBeforeProps
    }, (0, _getRenderPropValue.getRenderPropValue)(title)), content && /*#__PURE__*/_react.default.createElement("div", {
      className: contentCls
    }, (0, _getRenderPropValue.getRenderPropValue)(content)));
  }, [content, contentCls, cssScope, mode, title, titleBeforeProps]);
  return /*#__PURE__*/_react.default.createElement(_Tooltip.Tooltip, (0, _extends2.default)({
    ref: ref,
    prefixCls: cssScope(),
    overlay: popupNode,
    overlayClassName: (0, _clsx.default)(overlayClassName, cssScope(`--mode-${mode}`)),
    mouseLeaveDelay: mouseLeaveDelay,
    arrowHasStroke: true
  }, restProps));
});
if (process.env.NODE_ENV !== "production") {
  Popover.displayName = 'Popover';
}
