// @universe-design/colors/dist/index.esm.js
import { TinyColor } from '@ctrl/tinycolor';
export * from '@ctrl/tinycolor';
import _slicedToArray from '@babel/runtime/helpers/esm/slicedToArray';
import _createForOfIteratorHelper from '@babel/runtime/helpers/esm/createForOfIteratorHelper';
import _toConsumableArray from '@babel/runtime/helpers/esm/toConsumableArray';
import chroma from 'chroma-js';
import minBy from 'lodash/minBy';

/* eslint-disable @typescript-eslint/no-magic-numbers */
var ColorKeys;
(function (ColorKeys) {
  ColorKeys["Green"] = "green";
  ColorKeys["Blue"] = "blue";
  ColorKeys["Red"] = "red";
  ColorKeys["Orange"] = "orange";
  ColorKeys["Yellow"] = "yellow";
  ColorKeys["Sunflower"] = "sunflower";
  ColorKeys["Lime"] = "lime";
  ColorKeys["Turquoise"] = "turquoise";
  ColorKeys["Indigo"] = "indigo";
  ColorKeys["Purple"] = "purple";
  ColorKeys["Violet"] = "violet";
  ColorKeys["Carmine"] = "carmine";
  ColorKeys["Wathet"] = "wathet";
  ColorKeys["Neutral"] = "neutral";
})(ColorKeys || (ColorKeys = {}));
var DEFAULT_COLOR_LEVELS = [50, 100, 200, 300, 400, 500, 600, 700, 800, 900];

var PresetColors$1 = [{
  key: ColorKeys.Blue,
  colors: [{
    level: 50,
    value: '#f0f4ff'
  }, {
    level: 100,
    value: '#e1eaff'
  }, {
    level: 200,
    value: '#bacefd'
  }, {
    level: 300,
    value: '#82a7fc'
  }, {
    level: 400,
    value: '#4e83fd'
  }, {
    level: 500,
    value: '#3370ff'
  }, {
    level: 600,
    value: '#245bdb'
  }, {
    level: 700,
    value: '#1c4cba'
  }, {
    level: 800,
    value: '#133c9a'
  }, {
    level: 900,
    value: '#0c296e'
  }]
}, {
  key: ColorKeys.Carmine,
  colors: [{
    level: 50,
    value: '#fef1f8'
  }, {
    level: 100,
    value: '#fdddef'
  }, {
    level: 200,
    value: '#f9aed9'
  }, {
    level: 300,
    value: '#f57ac0'
  }, {
    level: 400,
    value: '#f14ba9'
  }, {
    level: 500,
    value: '#f01d94'
  }, {
    level: 600,
    value: '#c71077'
  }, {
    level: 700,
    value: '#9e1361'
  }, {
    level: 800,
    value: '#7a0f4b'
  }, {
    level: 900,
    value: '#520a32'
  }]
}, {
  key: ColorKeys.Green,
  colors: [{
    level: 50,
    value: '#f0fbef'
  }, {
    level: 100,
    value: '#d9f5d6'
  }, {
    level: 200,
    value: '#b7edb1'
  }, {
    level: 300,
    value: '#8ee085'
  }, {
    level: 400,
    value: '#62d256'
  }, {
    level: 500,
    value: '#34c724'
  }, {
    level: 600,
    value: '#2ea121'
  }, {
    level: 700,
    value: '#237b19'
  }, {
    level: 800,
    value: '#186010'
  }, {
    level: 900,
    value: '#124b0c'
  }]
}, {
  key: ColorKeys.Indigo,
  colors: [{
    level: 50,
    value: '#f2f3fd'
  }, {
    level: 100,
    value: '#e0e2fa'
  }, {
    level: 200,
    value: '#b4b9f3'
  }, {
    level: 300,
    value: '#7b83ea'
  }, {
    level: 400,
    value: '#616ae5'
  }, {
    level: 500,
    value: '#4954e6'
  }, {
    level: 600,
    value: '#2933c7'
  }, {
    level: 700,
    value: '#161fa2'
  }, {
    level: 800,
    value: '#0a1285'
  }, {
    level: 900,
    value: '#0c1264'
  }]
}, {
  key: ColorKeys.Lime,
  colors: [{
    level: 50,
    value: '#f7fae5'
  }, {
    level: 100,
    value: '#eef6c6'
  }, {
    level: 200,
    value: '#dfee96'
  }, {
    level: 300,
    value: '#d2e76a'
  }, {
    level: 400,
    value: '#c3dd40'
  }, {
    level: 500,
    value: '#b3d600'
  }, {
    level: 600,
    value: '#8fac02'
  }, {
    level: 700,
    value: '#667901'
  }, {
    level: 800,
    value: '#495700'
  }, {
    level: 900,
    value: '#354200'
  }]
}, {
  key: ColorKeys.Neutral,
  colors: [{
    level: 0,
    value: '#ffffff'
  }, {
    level: 50,
    value: '#f5f6f7'
  }, {
    level: 100,
    value: '#f2f3f5'
  }, {
    level: 200,
    value: '#eff0f1'
  }, {
    level: 300,
    value: '#dee0e3'
  }, {
    level: 400,
    value: '#bbbfc4'
  }, {
    level: 500,
    value: '#8f959e'
  }, {
    level: 600,
    value: '#646a73'
  }, {
    level: 650,
    value: '#51565d'
  }, {
    level: 700,
    value: '#373c43'
  }, {
    level: 800,
    value: '#2b2f36'
  }, {
    level: 900,
    value: '#1f2329'
  }, {
    level: 950,
    value: '#0f1114'
  }, {
    level: 1000,
    value: '#000000'
  }]
}, {
  key: ColorKeys.Orange,
  colors: [{
    level: 50,
    value: '#fff5eb'
  }, {
    level: 100,
    value: '#feead2'
  }, {
    level: 200,
    value: '#fed4a4'
  }, {
    level: 300,
    value: '#ffba6b'
  }, {
    level: 400,
    value: '#ffa53d'
  }, {
    level: 500,
    value: '#ff8800'
  }, {
    level: 600,
    value: '#de7802'
  }, {
    level: 700,
    value: '#b26206'
  }, {
    level: 800,
    value: '#8f4f04'
  }, {
    level: 900,
    value: '#6b3900'
  }]
}, {
  key: ColorKeys.Purple,
  colors: [{
    level: 50,
    value: '#f6f1fe'
  }, {
    level: 100,
    value: '#ece2fe'
  }, {
    level: 200,
    value: '#cdb2fa'
  }, {
    level: 300,
    value: '#ad82f7'
  }, {
    level: 400,
    value: '#935af6'
  }, {
    level: 500,
    value: '#7f3bf5'
  }, {
    level: 600,
    value: '#6425d0'
  }, {
    level: 700,
    value: '#4e1ba7'
  }, {
    level: 800,
    value: '#380d82'
  }, {
    level: 900,
    value: '#270561'
  }]
}, {
  key: ColorKeys.Red,
  colors: [{
    level: 50,
    value: '#fef1f1'
  }, {
    level: 100,
    value: '#fde2e2'
  }, {
    level: 200,
    value: '#fbbfbc'
  }, {
    level: 300,
    value: '#f98e8b'
  }, {
    level: 400,
    value: '#f76964'
  }, {
    level: 500,
    value: '#f54a45'
  }, {
    level: 600,
    value: '#d83931'
  }, {
    level: 700,
    value: '#ac2f28'
  }, {
    level: 800,
    value: '#812520'
  }, {
    level: 900,
    value: '#621c18'
  }]
}, {
  key: ColorKeys.Sunflower,
  colors: [{
    level: 50,
    value: '#fefff0'
  }, {
    level: 100,
    value: '#ffffcc'
  }, {
    level: 200,
    value: '#fffca3'
  }, {
    level: 300,
    value: '#fff67a'
  }, {
    level: 400,
    value: '#fff258'
  }, {
    level: 500,
    value: '#ffe928'
  }, {
    level: 600,
    value: '#e5ce00'
  }, {
    level: 700,
    value: '#c2a800'
  }, {
    level: 800,
    value: '#8f7c00'
  }, {
    level: 900,
    value: '#5c4f00'
  }]
}, {
  key: ColorKeys.Turquoise,
  colors: [{
    level: 50,
    value: '#eafaf8'
  }, {
    level: 100,
    value: '#d5f6f2'
  }, {
    level: 200,
    value: '#a9efe6'
  }, {
    level: 300,
    value: '#64e8d6'
  }, {
    level: 400,
    value: '#14e1c6'
  }, {
    level: 500,
    value: '#00d6b9'
  }, {
    level: 600,
    value: '#04b49c'
  }, {
    level: 700,
    value: '#078372'
  }, {
    level: 800,
    value: '#036356'
  }, {
    level: 900,
    value: '#024b41'
  }]
}, {
  key: ColorKeys.Violet,
  colors: [{
    level: 50,
    value: '#fceefc'
  }, {
    level: 100,
    value: '#f8def8'
  }, {
    level: 200,
    value: '#efb9ef'
  }, {
    level: 300,
    value: '#e58fe5'
  }, {
    level: 400,
    value: '#db66db'
  }, {
    level: 500,
    value: '#d136d1'
  }, {
    level: 600,
    value: '#b220b2'
  }, {
    level: 700,
    value: '#8c218c'
  }, {
    level: 800,
    value: '#681268'
  }, {
    level: 900,
    value: '#460b46'
  }]
}, {
  key: ColorKeys.Wathet,
  colors: [{
    level: 50,
    value: '#ecf9fe'
  }, {
    level: 100,
    value: '#d9f3fd'
  }, {
    level: 200,
    value: '#b1e8fc'
  }, {
    level: 300,
    value: '#7edafb'
  }, {
    level: 400,
    value: '#50cefb'
  }, {
    level: 500,
    value: '#14c0ff'
  }, {
    level: 600,
    value: '#049fd7'
  }, {
    level: 700,
    value: '#037eaa'
  }, {
    level: 800,
    value: '#006185'
  }, {
    level: 900,
    value: '#004761'
  }]
}, {
  key: ColorKeys.Yellow,
  colors: [{
    level: 50,
    value: '#fdf9ed'
  }, {
    level: 100,
    value: '#faf1d1'
  }, {
    level: 200,
    value: '#f8e6ab'
  }, {
    level: 300,
    value: '#f7dc82'
  }, {
    level: 400,
    value: '#fad355'
  }, {
    level: 500,
    value: '#ffc60a'
  }, {
    level: 600,
    value: '#dc9b04'
  }, {
    level: 700,
    value: '#aa7803'
  }, {
    level: 800,
    value: '#795101'
  }, {
    level: 900,
    value: '#5c3a00'
  }]
}];

var PresetColors = [{
  key: ColorKeys.Blue,
  colors: [{
    level: 50,
    value: '#151f33'
  }, {
    level: 100,
    value: '#192a4c'
  }, {
    level: 200,
    value: '#203e78'
  }, {
    level: 300,
    value: '#2851a3'
  }, {
    level: 400,
    value: '#2e65d1'
  }, {
    level: 500,
    value: '#4382ff'
  }, {
    level: 600,
    value: '#5e94ff'
  }, {
    level: 700,
    value: '#99bbff'
  }, {
    level: 800,
    value: '#b8cfff'
  }, {
    level: 900,
    value: '#cedfff'
  }]
}, {
  key: ColorKeys.Carmine,
  colors: [{
    level: 50,
    value: '#421b31'
  }, {
    level: 100,
    value: '#5c2542'
  }, {
    level: 200,
    value: '#91376a'
  }, {
    level: 300,
    value: '#ad407e'
  }, {
    level: 400,
    value: '#cc4792'
  }, {
    level: 500,
    value: '#e352a3'
  }, {
    level: 600,
    value: '#ed66b3'
  }, {
    level: 700,
    value: '#ff91cf'
  }, {
    level: 800,
    value: '#ffbde2'
  }, {
    level: 900,
    value: '#ffdbef'
  }]
}, {
  key: ColorKeys.Green,
  colors: [{
    level: 50,
    value: '#193016'
  }, {
    level: 100,
    value: '#1f471b'
  }, {
    level: 200,
    value: '#296b22'
  }, {
    level: 300,
    value: '#339428'
  }, {
    level: 400,
    value: '#3bb02e'
  }, {
    level: 500,
    value: '#55c948'
  }, {
    level: 600,
    value: '#6fd463'
  }, {
    level: 700,
    value: '#92de89'
  }, {
    level: 800,
    value: '#bdf0b8'
  }, {
    level: 900,
    value: '#d6f5d2'
  }]
}, {
  key: ColorKeys.Indigo,
  colors: [{
    level: 50,
    value: '#1d1f47'
  }, {
    level: 100,
    value: '#262a5e'
  }, {
    level: 200,
    value: '#353b8c'
  }, {
    level: 300,
    value: '#4149b5'
  }, {
    level: 400,
    value: '#4c55d4'
  }, {
    level: 500,
    value: '#5f69ed'
  }, {
    level: 600,
    value: '#737beb'
  }, {
    level: 700,
    value: '#9ea5ff'
  }, {
    level: 800,
    value: '#bbc0f6'
  }, {
    level: 900,
    value: '#d2d5fd'
  }]
}, {
  key: ColorKeys.Lime,
  colors: [{
    level: 50,
    value: '#2e3605'
  }, {
    level: 100,
    value: '#495706'
  }, {
    level: 200,
    value: '#6d8205'
  }, {
    level: 300,
    value: '#91ad02'
  }, {
    level: 400,
    value: '#abc912'
  }, {
    level: 500,
    value: '#b8d911'
  }, {
    level: 600,
    value: '#c7e04d'
  }, {
    level: 700,
    value: '#d6ea75'
  }, {
    level: 800,
    value: '#e7f5a6'
  }, {
    level: 900,
    value: '#f6ffcc'
  }]
}, {
  key: ColorKeys.Neutral,
  colors: [{
    level: 0,
    value: '#101010'
  }, {
    level: 50,
    value: '#1f1f1f'
  }, {
    level: 100,
    value: '#2e2e2e'
  }, {
    level: 200,
    value: '#373737'
  }, {
    level: 300,
    value: '#434343'
  }, {
    level: 400,
    value: '#5f5f5f'
  }, {
    level: 500,
    value: '#939393'
  }, {
    level: 600,
    value: '#c1c1c1'
  }, {
    level: 650,
    value: '#cfcfcf'
  }, {
    level: 700,
    value: '#e0e0e0'
  }, {
    level: 800,
    value: '#e8e8e8'
  }, {
    level: 900,
    value: '#f0f0f0'
  }, {
    level: 950,
    value: '#f8f8f8'
  }, {
    level: 1000,
    value: '#ffffff'
  }]
}, {
  key: ColorKeys.Orange,
  colors: [{
    level: 50,
    value: '#33210b'
  }, {
    level: 100,
    value: '#57330a'
  }, {
    level: 200,
    value: '#845117'
  }, {
    level: 300,
    value: '#bd7017'
  }, {
    level: 400,
    value: '#e58312'
  }, {
    level: 500,
    value: '#ff9e2e'
  }, {
    level: 600,
    value: '#ffac4d'
  }, {
    level: 700,
    value: '#ffc077'
  }, {
    level: 800,
    value: '#ffd8ac'
  }, {
    level: 900,
    value: '#ffeace'
  }]
}, {
  key: ColorKeys.Purple,
  colors: [{
    level: 50,
    value: '#281745'
  }, {
    level: 100,
    value: '#361d61'
  }, {
    level: 200,
    value: '#4d2691'
  }, {
    level: 300,
    value: '#6533bd'
  }, {
    level: 400,
    value: '#7c4ad4'
  }, {
    level: 500,
    value: '#9762f5'
  }, {
    level: 600,
    value: '#a472fc'
  }, {
    level: 700,
    value: '#b990ff'
  }, {
    level: 800,
    value: '#d3bafd'
  }, {
    level: 900,
    value: '#e3d4ff'
  }]
}, {
  key: ColorKeys.Red,
  colors: [{
    level: 50,
    value: '#361918'
  }, {
    level: 100,
    value: '#4a1d1b'
  }, {
    level: 200,
    value: '#6b2827'
  }, {
    level: 300,
    value: '#9e3836'
  }, {
    level: 400,
    value: '#cc4743'
  }, {
    level: 500,
    value: '#f0544f'
  }, {
    level: 600,
    value: '#fa7873'
  }, {
    level: 700,
    value: '#ff9c99'
  }, {
    level: 800,
    value: '#fec6c3'
  }, {
    level: 900,
    value: '#ffe0e0'
  }]
}, {
  key: ColorKeys.Sunflower,
  colors: [{
    level: 50,
    value: '#38340c'
  }, {
    level: 100,
    value: '#5c5514'
  }, {
    level: 200,
    value: '#877b10'
  }, {
    level: 300,
    value: '#c7b300'
  }, {
    level: 400,
    value: '#e8d100'
  }, {
    level: 500,
    value: '#f7e328'
  }, {
    level: 600,
    value: '#faed50'
  }, {
    level: 700,
    value: '#fcf47c'
  }, {
    level: 800,
    value: '#fffc9e'
  }, {
    level: 900,
    value: '#ffffc7'
  }]
}, {
  key: ColorKeys.Turquoise,
  colors: [{
    level: 50,
    value: '#162e2b'
  }, {
    level: 100,
    value: '#153d38'
  }, {
    level: 200,
    value: '#23665d'
  }, {
    level: 300,
    value: '#269687'
  }, {
    level: 400,
    value: '#2ab8a4'
  }, {
    level: 500,
    value: '#34d9c3'
  }, {
    level: 600,
    value: '#49e5d0'
  }, {
    level: 700,
    value: '#7febdc'
  }, {
    level: 800,
    value: '#b7f7ef'
  }, {
    level: 900,
    value: '#daf9f5'
  }]
}, {
  key: ColorKeys.Violet,
  colors: [{
    level: 50,
    value: '#3b153b'
  }, {
    level: 100,
    value: '#4f1c4f'
  }, {
    level: 200,
    value: '#7a287a'
  }, {
    level: 300,
    value: '#943b94'
  }, {
    level: 400,
    value: '#b54ab5'
  }, {
    level: 500,
    value: '#d457d4'
  }, {
    level: 600,
    value: '#de73de'
  }, {
    level: 700,
    value: '#e898e8'
  }, {
    level: 800,
    value: '#f2bff2'
  }, {
    level: 900,
    value: '#fbdafb'
  }]
}, {
  key: ColorKeys.Wathet,
  colors: [{
    level: 50,
    value: '#142930'
  }, {
    level: 100,
    value: '#173742'
  }, {
    level: 200,
    value: '#225b70'
  }, {
    level: 300,
    value: '#2782a3'
  }, {
    level: 400,
    value: '#2ea5d1'
  }, {
    level: 500,
    value: '#42bdeb'
  }, {
    level: 600,
    value: '#5fd4fe'
  }, {
    level: 700,
    value: '#89dffe'
  }, {
    level: 800,
    value: '#c2eeff'
  }, {
    level: 900,
    value: '#def6ff'
  }]
}, {
  key: ColorKeys.Yellow,
  colors: [{
    level: 50,
    value: '#382e0c'
  }, {
    level: 100,
    value: '#574711'
  }, {
    level: 200,
    value: '#8f7211'
  }, {
    level: 300,
    value: '#c99d08'
  }, {
    level: 400,
    value: '#f0bb0a'
  }, {
    level: 500,
    value: '#ffcb1f'
  }, {
    level: 600,
    value: '#ffd759'
  }, {
    level: 700,
    value: '#ffde75'
  }, {
    level: 800,
    value: '#ffe694'
  }, {
    level: 900,
    value: '#fcefc2'
  }]
}];

// const hueStep = 2; // 色相阶梯
// const saturationStep = 0.16; // 饱和度阶梯，浅色部分
// const saturationStep2 = 0.05; // 饱和度阶梯，深色部分
// const brightnessStep1 = 0.05; // 亮度阶梯，浅色部分
var brightnessStep2 = 0.15; // 亮度阶梯，深色部分
// const lightColorCount = 5; // 浅色数量，主色上
// const darkColorCount = 4; // 深色数量，主色下
var hueStep = 0; // 色相阶梯
var saturationStep = 0.17; // 饱和度阶梯，浅色部分
var saturationStep2 = 0.05; // 饱和度阶梯，深色部分
// const brightnessStep1 = 0.04; // 亮度阶梯，浅色部分
// const brightnessStep2 = 0.14; // 亮度阶梯，深色部分
var lightColorCount = 5; // 浅色数量，主色上
var darkColorCount = 4; // 深色数量，主色下
function getHue(hsv, i, light) {
  var hue;
  // 根据色相不同，色相转向不同
  if (Math.round(hsv.h) >= 60 && Math.round(hsv.h) <= 230) {
    hue = light ? Math.round(hsv.h) - hueStep * i : Math.round(hsv.h) + hueStep * i;
  } else {
    hue = light ? Math.round(hsv.h) + hueStep * i : Math.round(hsv.h) - hueStep * i;
  }
  if (hue < 0) {
    hue += 360;
  } else if (hue >= 360) {
    hue -= 360;
  }
  return hue;
}
function bezier3(t, controllPoints) {
  var calc = function () {
    return Math.pow(1 - t, 3) * (arguments.length <= 0 ? undefined : arguments[0]) + 3 * t * Math.pow(1 - t, 2) * (arguments.length <= 1 ? undefined : arguments[1]) + 3 * Math.pow(t, 2) * (1 - t) * (arguments.length <= 2 ? undefined : arguments[2]) + Math.pow(t, 3) * (arguments.length <= 3 ? undefined : arguments[3]);
  };
  return {
    x: calc.apply(void 0, _toConsumableArray(controllPoints.map(function (o) {
      return o.x;
    }))),
    y: calc.apply(void 0, _toConsumableArray(controllPoints.map(function (o) {
      return o.y;
    })))
  };
}
function getSaturation(hsv, i, light) {
  var MIN = 0.05;
  // grey color don't change saturation
  if (hsv.h === 0 && hsv.s === 0) {
    return hsv.s;
  }
  var saturation;
  if (light) {
    // 0 阶
    // saturation = hsv.s - saturationStep * i;
    // 1 阶
    // const step = (hsv.s - MIN) / lightColorCount;
    // saturation = hsv.s - step * i;
    // 2 阶
    // const t = (lightColorCount - i) / lightColorCount;
    // const controllPoint = (hsv.s + MIN) / 6; // 经过点
    // saturation =
    //   (1 - t) ** 2 * MIN + 2 * t * (1 - t) * controllPoint + t ** 2 * hsv.s;
    // 3 阶

    saturation = bezier3((lightColorCount - i) / lightColorCount, [{
      x: 0,
      y: MIN
    }, {
      x: 1,
      y: (hsv.s - MIN) * 0.05 + MIN
    }, {
      x: 4,
      y: (hsv.s - MIN) * 0.8 + MIN
    }, {
      x: 5,
      y: hsv.s
    }]).y;
  } else if (i === darkColorCount) {
    saturation = hsv.s + saturationStep;
  } else {
    saturation = hsv.s + saturationStep2 * i;
  }
  // 边界值修正
  if (saturation > 1) {
    saturation = 1;
  }
  // 第一格的 s 限制在 0.06-0.1 之间
  if (light && i === lightColorCount && saturation > 0.1) {
    saturation = 0.1;
  }
  if (saturation < MIN) {
    saturation = MIN;
  }
  return Number(saturation.toFixed(2));
}
// function getSaturation2(hsv: HsvObject, i: number, light?: boolean): number {
//   return
// }
function getValue(color, i, light) {
  var MAX = 1;
  var hsv = color.toHsv();
  // const greyScale = color.toHsl().l;
  var value;
  if (light) {
    // value = hsv.v + brightnessStep1 * i;
    var step = (MAX - hsv.v) / lightColorCount;
    value = hsv.v + step * i;
  } else {
    value = hsv.v - brightnessStep2 * i;
    // 3 阶
    // console.log((hsv.s + 0.1) ** 8);
    // const t = i / darkColorCount;
    // const MIN = hsv.v - 0.57;
    // value = bezier3(t, [
    //   { x: 0, y: hsv.v },
    //   { x: 1, y: ((hsv.v - MIN) * 0.1) / (hsv.s + 0.1) ** 16 + MIN }, // (hsv.s + 0.1) ** 16 为了让 s 大的下降快一些，s 小的下降慢一点
    //   { x: 3, y: ((hsv.v - MIN) * 0.1) / (hsv.s + 0.1) ** 16 + MIN },
    //   { x: 4, y: MIN },
    // ]).y;
  }
  if (value > MAX) {
    value = MAX;
  }
  // console.log(value);
  // if (value < 0.06) {
  //   value = 0.06;
  // }
  return Number(value.toFixed(2));
}
function generateIsolatedColor_lark (color) {
  var patterns = [];
  var pColor = new TinyColor(color);
  var _iterator = _createForOfIteratorHelper(DEFAULT_COLOR_LEVELS.entries()),
    _step;
  try {
    for (_iterator.s(); !(_step = _iterator.n()).done;) {
      var _step$value = _slicedToArray(_step.value, 2),
        i = _step$value[0],
        level = _step$value[1];
      var isLight = i <= lightColorCount;
      var hsv = pColor.toHsv();
      var colorString = new TinyColor({
        h: getHue(hsv, isLight ? lightColorCount - i : i - lightColorCount, isLight),
        s: getSaturation(hsv, isLight ? lightColorCount - i : i - lightColorCount, isLight),
        v: getValue(pColor, isLight ? lightColorCount - i : i - lightColorCount, isLight)
      }).toHexString();
      patterns.push({
        level: level,
        value: colorString
      });
    }
  } catch (err) {
    _iterator.e(err);
  } finally {
    _iterator.f();
  }
  return patterns;
}

var between = function (a, min, max) {
  return Math.min(Math.max(a, min), max);
};
var presetColors = PresetColors$1;
function generateFullColorPalette_lark () {
  var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
    _ref$brandColor = _ref.brandColor,
    brandColor = _ref$brandColor === void 0 ? PresetColors$1[0].colors[6].value : _ref$brandColor;
  var chromaBrandColor = chroma(brandColor);
  var chromaColors = presetColors.map(function (o) {
    return o.colors[6].value;
  }).flat().map(function (o) {
    return chroma(o);
  });
  var closestColor = minBy(chromaColors, function (c) {
    return chroma.deltaE(c, chromaBrandColor);
  });
  if (!closestColor) {
    return presetColors;
  }
  var _closestColor$hsv = closestColor.hsv(),
    _closestColor$hsv2 = _slicedToArray(_closestColor$hsv, 3),
    h1 = _closestColor$hsv2[0],
    s1 = _closestColor$hsv2[1],
    v1 = _closestColor$hsv2[2];
  var _chromaBrandColor$hsv = chromaBrandColor.hsv(),
    _chromaBrandColor$hsv2 = _slicedToArray(_chromaBrandColor$hsv, 3),
    h2 = _chromaBrandColor$hsv2[0],
    s2 = _chromaBrandColor$hsv2[1],
    v2 = _chromaBrandColor$hsv2[2];
  var _ref2 = [between(h2 - h1, -10, 10), between(s2 - s1, -0.1, 0.1), between(v2 - v1, -0.1, 0.1)],
    dh = _ref2[0],
    ds = _ref2[1],
    dv = _ref2[2];
  var getNewColor = function (color) {
    var chromaColor = chroma(color);
    var _chromaColor$hsv = chromaColor.hsv(),
      _chromaColor$hsv2 = _slicedToArray(_chromaColor$hsv, 3),
      h = _chromaColor$hsv2[0],
      s = _chromaColor$hsv2[1],
      v = _chromaColor$hsv2[2];
    var newColor = chromaColor.set('hsv.h', (h + dh + 360) % 360).set('hsv.s', s + ds).set('hsv.v', v + dv);
    return newColor;
  };
  return presetColors.map(function (_ref3) {
    var key = _ref3.key,
      colors = _ref3.colors;
    var newColors = colors.map(function (_ref4) {
      var level = _ref4.level,
        value = _ref4.value;
      return {
        level: level,
        value: getNewColor(value).hex()
      };
    });
    return {
      key: key,
      colors: newColors
    };
  });
}

// 暗色主题颜色映射关系表
var darkColorMap$1 = [{
  index: 9,
  s: 0.92,
  v: -0.1,
  level: 50
}, {
  index: 9,
  s: 0.92,
  v: -0.05,
  level: 100
}, {
  index: 8,
  s: 0.92,
  v: -0.05,
  level: 200
}, {
  index: 7,
  s: 0.92,
  v: -0.05,
  level: 300
}, {
  index: 6,
  s: 0.92,
  v: 0.01,
  level: 400
}, {
  index: 5,
  s: 0.92,
  v: 0.01,
  level: 500
}, {
  index: 4,
  s: 0.92,
  v: 0.01,
  level: 600
}, {
  index: 3,
  s: 0.92,
  v: 0.01,
  level: 700
}, {
  index: 2,
  s: 0.92,
  v: 0.01,
  level: 800
}, {
  index: 1,
  s: 0.92,
  v: 0.01,
  level: 900
}];
var generateIsolatedDarkColor_lark = (function (lightColors) {
  return darkColorMap$1.map(function (_ref) {
    var index = _ref.index,
      sMount = _ref.s,
      vMount = _ref.v,
      level = _ref.level;
    var value = lightColors[index].value;
    var _TinyColor$toHsv = new TinyColor(value).toHsv(),
      h = _TinyColor$toHsv.h,
      s = _TinyColor$toHsv.s,
      v = _TinyColor$toHsv.v;
    var darkColorString = new TinyColor({
      h: h,
      s: s * sMount,
      v: Math.max(Math.min(v + vMount, 1), 0)
    }).toHexString();
    return {
      level: level,
      value: darkColorString
    };
  });
});

// 暗色主题颜色映射关系表
var darkColorMap = [{
  index: 7,
  opacity: 0.15,
  level: 50
}, {
  index: 6,
  opacity: 0.25,
  level: 100
}, {
  index: 5,
  opacity: 0.3,
  level: 200
}, {
  index: 5,
  opacity: 0.45,
  level: 300
}, {
  index: 5,
  opacity: 0.65,
  level: 400
}, {
  index: 5,
  opacity: 0.85,
  level: 500
}, {
  index: 4,
  opacity: 0.9,
  level: 600
}, {
  index: 3,
  opacity: 0.95,
  level: 700
}, {
  index: 2,
  opacity: 0.97,
  level: 800
}, {
  index: 1,
  opacity: 0.98,
  level: 900
}];
var generateIsolatedDarkColor_antd = (function (lightColors, props) {
  return darkColorMap.map(function (_ref) {
    var index = _ref.index,
      opacity = _ref.opacity,
      level = _ref.level;
    var value = lightColors[index].value;
    var darkColorString = new TinyColor(props.backgroundColor).mix(value, opacity * 100).toHexString();
    return {
      level: level,
      value: darkColorString
    };
  });
});

export { ColorKeys, DEFAULT_COLOR_LEVELS, PresetColors as PresetColorsLarkDark, PresetColors$1 as PresetColorsLarkLight, generateFullColorPalette_lark as generateFullColorPalette, generateIsolatedColor_lark as generateIsolatedColor, generateIsolatedDarkColor_lark as generateIsolatedDarkColor, generateIsolatedDarkColor_antd as generateIsolatedDarkColorAntd };
