// @universe-design/colors/dist/index.js
'use strict';

var tinycolor = require('@ctrl/tinycolor');
var light = require('./preset-colors/light.js');
var dark = require('./preset-colors/dark.js');
var generateIsolatedColor_lark = require('./algorithm/generate-isolated-color.lark.js');
var generateFullColorPalette_lark = require('./algorithm/generate-full-color-palette.lark.js');
var generateIsolatedDarkColor_lark = require('./algorithm/generate-isolated-dark-color.lark.js');
var generateIsolatedDarkColor_antd = require('./algorithm/generate-isolated-dark-color.antd.js');
var types = require('./types.js');



exports.PresetColorsLarkLight = light.default;
exports.PresetColorsLarkDark = dark.default;
exports.generateIsolatedColor = generateIsolatedColor_lark.default;
exports.generateFullColorPalette = generateFullColorPalette_lark.default;
exports.generateIsolatedDarkColor = generateIsolatedDarkColor_lark.default;
exports.generateIsolatedDarkColorAntd = generateIsolatedDarkColor_antd.default;
Object.defineProperty(exports, 'ColorKeys', {
	enumerable: true,
	get: function () { return types.ColorKeys; }
});
exports.DEFAULT_COLOR_LEVELS = types.DEFAULT_COLOR_LEVELS;
Object.keys(tinycolor).forEach(function (k) {
	if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
		enumerable: true,
		get: function () { return tinycolor[k]; }
	});
});
