// @universe-design/react/lib/shared/components/popper/imperative.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.popper = exports.ImperativePopperController = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var React = _interopRequireWildcard(require("react"));
var ReactDOM = _interopRequireWildcard(require("react-dom"));
var _Popper = require("./Popper");
var _configProvider = require("../../../config-provider");
var _util = require("../../../util");
const ImperativePopperReference = /*#__PURE__*/React.forwardRef((props, ref) => {
  React.useImperativeHandle(ref, () => props.reference);
  return /*#__PURE__*/React.createElement(React.Fragment, null);
});
const ImperativePopper = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    defaultProps
  } = _ref;
  const [props, setProps] = React.useState(defaultProps);
  React.useImperativeHandle(ref, () => ({
    update: nextProps => {
      setProps(nextProps);
    }
  }), []);
  return /*#__PURE__*/React.createElement(_Popper.Popper, props);
});
class ImperativePopperController {
  constructor(_props, popper) {
    (0, _defineProperty2.default)(this, "_popperManager", void 0);
    // 用做 ReactDOM render 的 container
    (0, _defineProperty2.default)(this, "_container", void 0);
    (0, _defineProperty2.default)(this, "_imperativePopperRef", /*#__PURE__*/React.createRef());
    (0, _defineProperty2.default)(this, "_isAlive", false);
    (0, _defineProperty2.default)(this, "update", props => {
      this._initOrUpdate(props);
    });
    (0, _defineProperty2.default)(this, "destroy", () => {
      this._cleaup();
    });
    (0, _defineProperty2.default)(this, "_initOrUpdate", props => {
      var _this$_imperativePopp;
      if (!this._isAlive) {
        // console.log('_init', this, props);
        this._isAlive = true;
        this._container = document.createElement('div');
        document.body.appendChild(this._container);
        ReactDOM.render(/*#__PURE__*/React.createElement(_configProvider.CrossRootConfigProvider, null, /*#__PURE__*/React.createElement(ImperativePopper, {
          ref: this._imperativePopperRef,
          defaultProps: this.props2PopperProps(props)
        })), this._container);
        this._popperManager.controllers.add(this);
        return;
      }
      // console.log('_update', this, props);
      (_this$_imperativePopp = this._imperativePopperRef.current) === null || _this$_imperativePopp === void 0 ? void 0 : _this$_imperativePopp.update(this.props2PopperProps(props));
    });
    (0, _defineProperty2.default)(this, "_cleaup", () => {
      this._popperManager.controllers.delete(this);
      if (this._container) {
        ReactDOM.unmountComponentAtNode(this._container);
        document.body.removeChild(this._container);
      }
      this._isAlive = false;
    });
    this._popperManager = popper;
    this._initOrUpdate(_props);
  }
  props2PopperProps(props) {
    const {
      reference,
      destroyOnHide = true,
      onAfterClose,
      ...popperProps
    } = props;
    return {
      // defaultVisible: true,
      popupPortalContainer: false,
      // 默认在 container 原位渲染
      ...popperProps,
      destroyOnHide,
      onAfterClose: () => {
        (0, _util.isFunction)(onAfterClose) && onAfterClose();
        if (destroyOnHide) {
          this._cleaup();
        }
      },
      children: /*#__PURE__*/React.createElement(ImperativePopperReference, {
        reference: reference
      }),
      domOnly: false // 为了让 virtual element 生效需要使用 domOnly false
    };
  }
}
exports.ImperativePopperController = ImperativePopperController;
class PopperManager {
  constructor() {
    (0, _defineProperty2.default)(this, "controllers", new Set());
    (0, _defineProperty2.default)(this, "create", props => new ImperativePopperController(props, this));
  }
}
const popper = exports.popper = new PopperManager();
