// @universe-design/react/lib/shared/components/popper/PopperPopupVisibleController.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PopperPopupVisibleController = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactTransitionGroup = require("react-transition-group");
var _useAnimatedVisible = require("../transition/useAnimatedVisible");
var _types = require("./types");
var _util = require("../../../util");
var _PopperContext = require("./PopperContext");
const PopperPopupVisibleController = props => {
  const {
    timeout = _types.defaults.timeout,
    forceRender = false,
    destroyOnHide = true,
    onAfterClose,
    onAfterOpen,
    children
  } = props;
  const {
    visible,
    popperLatestValues
  } = (0, _PopperContext.usePopperContext)();
  const noopRef = (0, _useAnimatedVisible.useTransitionGroupNoopNodeRef)();
  return /*#__PURE__*/_react.default.createElement(_reactTransitionGroup.Transition, {
    nodeRef: noopRef,
    timeout: timeout,
    appear: true,
    in: visible,
    mountOnEnter: !forceRender,
    unmountOnExit: destroyOnHide,
    onEnter: () => {
      var _popperLatestValues$u;
      (_popperLatestValues$u = popperLatestValues.update) === null || _popperLatestValues$u === void 0 ? void 0 : _popperLatestValues$u.call(popperLatestValues);
    },
    onEntered: () => {
      (0, _util.isFunction)(onAfterOpen) && onAfterOpen();
    },
    onExited: () => {
      var _popperLatestValues$r;
      (0, _util.isFunction)(onAfterClose) && onAfterClose();
      (_popperLatestValues$r = popperLatestValues.resetPlacement) === null || _popperLatestValues$r === void 0 ? void 0 : _popperLatestValues$r.call(popperLatestValues);
    }
  }, children);
};
exports.PopperPopupVisibleController = PopperPopupVisibleController;
