// @universe-design/react/lib/shared/components/popper/PopperProvider.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PopperProvider = void 0;
var React = _interopRequireWildcard(require("react"));
var _reactUtilities = require("@universe-design/react-utilities");
var _composeRefs = _interopRequireDefault(require("../../../util/composeRefs"));
var _usePopper = require("./usePopper");
var _usePopperAction = require("./usePopperAction");
var _util = require("../../../util");
var _dom = require("../../../util/dom");
var _configProvider = require("../../../config-provider");
var _PopperContext = require("./PopperContext");
const PopperProvider = exports.PopperProvider = /*#__PURE__*/React.forwardRef((props, ref) => {
  const {
    visible: visibleProp,
    checkOverflow = false,
    defaultVisible = false,
    children,
    trigger: triggerProp,
    showAction: showActionProp,
    hideAction: hideActionProp,
    mouseEnterDelay,
    mouseLeaveDelay: mouseLeaveDelayProp,
    onVisibleChange,
    udTriggerRef,
    disabled = false,
    prefixCls,
    modifiers,
    __screenshotAlwaysOpenEnabled = false
  } = props;

  /**
   * Deprecated Warning
   */
  process.env.NODE_ENV !== "production" ? (0, _util.warning)(!udTriggerRef, (0, _util.deprecatedWarning)('Popper', 'udTriggerRef', 'ref')) : void 0;
  process.env.NODE_ENV !== "production" ? (0, _util.warning)(!modifiers, (0, _util.deprecatedWarning)('Popper', 'modifiers', ['eventListeners', 'enableMaxHeight'])) : void 0;
  const rootPrefixCls = (0, _configProvider.useCssScope)('popper', prefixCls)('');

  /**
   * Popper 打开状态
   */
  const [visible, _setVisible] = (0, _reactUtilities.useControlledState)(defaultVisible, visibleProp, {
    onChange: onVisibleChange
  });

  /**
   * inline
   */
  const inlineElementMousePositionRef = React.useRef({
    x: 0,
    y: 0
  });

  /**
   * Popper 对齐
   */
  const {
    referenceRef,
    referenceElementRef,
    referenceElement,
    popperLatestValues
  } = (0, _usePopper.usePopperShared)();
  const checkVisible = React.useCallback(_visible => {
    if (_visible && checkOverflow) {
      const overflowChecker = typeof checkOverflow === 'function' ? checkOverflow : _dom.isOverflowed;
      const child = referenceElementRef.current;
      if (!child || !overflowChecker(child)) {
        _setVisible(false);
        return;
      }
    }
    _setVisible(_visible);
  }, [checkOverflow, _setVisible, referenceElementRef]);

  // 保证在截图时一定能弹出
  _util.__SCREENSHOT__ && __screenshotAlwaysOpenEnabled &&
  // eslint-disable-next-line react-hooks/rules-of-hooks
  React.useLayoutEffect(() => {
    checkVisible(true);
  });

  /**
   * Popper 触发方式
   */
  const {
    handleReferenceClick,
    handleContextMenu,
    handleMouseEnter,
    handleMouseLeave,
    handleKeyDown,
    handleReferenceFocus,
    setVisible,
    showAction,
    hideAction,
    alignPoint,
    mouseLeaveDelay,
    setVisibleDelay
  } = (0, _usePopperAction.usePopperActionShared)({
    trigger: triggerProp,
    showAction: showActionProp,
    hideAction: hideActionProp,
    mouseEnterDelay,
    mouseLeaveDelay: mouseLeaveDelayProp
  }, {
    visible,
    setVisible: checkVisible,
    referenceRef,
    referenceElementRef,
    inlineElementMousePositionRef
  });
  const mergedPopperRef = React.useMemo(() => (0, _composeRefs.default)(udTriggerRef, ref), [ref, udTriggerRef]);
  React.useImperativeHandle(mergedPopperRef, () => ({
    visible,
    setVisible,
    forcePopupAlign: () => {
      var _popperLatestValues$u;
      return (_popperLatestValues$u = popperLatestValues.update) === null || _popperLatestValues$u === void 0 ? void 0 : _popperLatestValues$u.call(popperLatestValues);
    }
  }), [popperLatestValues, setVisible, visible]);
  return /*#__PURE__*/React.createElement(_PopperContext.PopperContextProvider, {
    value: {
      showAction,
      hideAction,
      alignPoint,
      visible,
      handleReferenceClick,
      handleContextMenu,
      handleMouseEnter,
      handleMouseLeave,
      handleKeyDown,
      handleReferenceFocus,
      setVisible,
      setVisibleDelay,
      referenceRef,
      referenceElement,
      referenceElementRef,
      popperLatestValues,
      disabled,
      rootPrefixCls,
      mouseLeaveDelay,
      inlineElementMousePositionRef
    }
  }, children);
});
if (process.env.NODE_ENV !== "production") {
  PopperProvider.displayName = 'PopperProvider';
}
