// @universe-design/react/lib/shared/components/popper/PopperReference.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PopperReference = void 0;
var React = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _reactUtilities = require("@universe-design/react-utilities");
var _PopperContext = require("./PopperContext");
var _util = require("../../../util");
var _function = require("../../../util/function");
var _hooks = require("../../../hooks");
const PopperReference = props => {
  const {
    rootPrefixCls,
    referenceRef,
    handleReferenceClick,
    handleMouseEnter,
    handleMouseLeave,
    handleContextMenu,
    handleKeyDown,
    handleReferenceFocus,
    alignPoint,
    disabled,
    visible
  } = (0, _PopperContext.usePopperContext)();
  const {
    children: childrenProp,
    className,
    style,
    openClassName = `${rootPrefixCls}-open`,
    domOnly = true
  } = props;

  // resolve function children
  let children = (0, _util.isFunction)(childrenProp) ? childrenProp({
    visible
  }) : childrenProp;
  const mergedClassName = (0, _clsx.default)(children.props.className, className, visible && openClassName);
  const eventProps = {
    onClick: (0, _hooks.useEventCallback)((0, _function.composeEventHandlers)(children.props.onClick, props.onClick, handleReferenceClick)),
    onMouseEnter: (0, _hooks.useEventCallback)((0, _function.composeEventHandlers)(children.props.onMouseEnter, props.onMouseEnter, handleMouseEnter)),
    onMouseLeave: (0, _hooks.useEventCallback)((0, _function.composeEventHandlers)(children.props.onMouseLeave, props.onMouseLeave, handleMouseLeave)),
    onFocus: (0, _hooks.useEventCallback)((0, _function.composeEventHandlers)(children.props.onFocus, props.onFocus, handleReferenceFocus)),
    onContextMenu: (0, _hooks.useEventCallback)((0, _function.composeEventHandlers)(children.props.onContextMenu, props.onContextMenu, handleContextMenu)),
    onKeyDown: (0, _hooks.useEventCallback)((0, _function.composeEventHandlers)(children.props.onKeyDown, props.onKeyDown, handleKeyDown))
  };

  // clone props
  children = /*#__PURE__*/React.cloneElement(children, {
    ...(mergedClassName ? {
      className: mergedClassName
    } : {}),
    style: {
      ...style,
      ...children.props.style
    },
    ...(visible && !(0, _util.isNonNullable)(children.props['aria-expanded']) ? {
      'aria-expanded': visible
    } : {}),
    ...(disabled ? {} : eventProps)
  });

  // clone ref
  children = (0, _reactUtilities.useCloneElementRef)({
    children,
    ref: referenceRef
  }, {
    skipped: alignPoint || disabled,
    domOnly
  });
  return children;
};
exports.PopperReference = PopperReference;
