// @universe-design/react/lib/shared/components/popper/PopperPopup.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PopperPopup = exports.PopperArrow = void 0;
var _react = _interopRequireDefault(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _portal = _interopRequireDefault(require("../../../portal"));
var _ZoomTransition = require("../transition/ZoomTransition");
var _StreamlineArrow = require("../trigger/StreamlineArrow");
var _PopperDrillingContext = require("./PopperDrillingContext");
var _FocusLock = require("../FocusLock");
var _util = require("../../../util");
var _usePopperAction = require("./usePopperAction");
var _usePopper = require("./usePopper");
var _PopperPopupContext = require("./PopperPopupContext");
var _PopperContext = require("./PopperContext");
var _types = require("./types");
var _reactUtilities = require("@universe-design/react-utilities");
var _reactProvider = require("@universe-design/react-provider");
// import { useLogger } from '../../../hooks';

const PopperArrow = _ref => {
  let {
    hasStroke
  } = _ref;
  const {
    rootPrefixCls
  } = (0, _PopperContext.usePopperContext)();
  const {
    placement
  } = (0, _PopperPopupContext.usePopperPopupContext)();
  const direction = (() => {
    if (!placement) {
      return undefined;
    }
    if (placement.startsWith('right') || placement.startsWith('left')) {
      return 'vertical';
    }
    if (placement.startsWith('top') || placement.startsWith('bottom')) {
      return 'horizontal';
    }
  })();
  return /*#__PURE__*/_react.default.createElement(_StreamlineArrow.StreamlineArrow, {
    direction: direction,
    className: `${rootPrefixCls}__arrow__streamline`,
    hasStroke: hasStroke,
    style: {
      display: 'flex',
      transform: (() => {
        // 由于 arrow 尺寸跟 placement 有关系，所以不能占 arrow container 的宽度，不然定位的时候是 arrow 大小是不对的，定位以后 arrow 更新会歪掉，但是这样写就意味着 arrowPadding 要手动加上 arrow 一半的宽度
        if (!placement) {
          return;
        }
        if (placement.startsWith('right')) {
          return `translate(${hasStroke ? 'calc(1px - 100%)' : '-100%'}, -50%) rotate(180deg)`;
        }
        if (placement.startsWith('left')) {
          return `translate(${hasStroke ? '-1px' : '0'}, -50%)`;
        }
        if (placement.startsWith('top')) {
          return `translate(-50%, ${hasStroke ? '-1px' : '0'})`;
        }
        if (placement.startsWith('bottom')) {
          return `translate(-50%, ${hasStroke ? 'calc(1px - 100%)' : '-100%'}) rotate(180deg)`;
        }
      })(),
      zIndex: 'inherit',
      position: 'absolute'
    }
  });
};
exports.PopperArrow = PopperArrow;
const PopperPopup = props => {
  const {
    children,
    Transition: UserTransition = _ZoomTransition.ZoomTransition,
    timeout = _types.defaults.timeout,
    arrow,
    popupPortalContainer,
    getPopupContainer,
    popupClassName,
    popupStyle,
    initialFocus,
    enableFocusLock = false,
    enableRestoreFocus = true,
    inline = false,
    transitionStatus: status,
    popupStopPropagation = {},
    ...popperProps
  } = props;
  const {
    click: popupClickStopPropagation = false
  } = popupStopPropagation;

  /**
   * Deprecated Warning
   */
  process.env.NODE_ENV !== "production" ? (0, _util.warning)(!getPopupContainer, (0, _util.deprecatedWarning)('Popper', 'getPopupContainer', 'popupPortalContainer')) : void 0;
  const {
    handleMouseEnter,
    handleMouseLeave,
    handleKeyDown,
    visible,
    referenceElementRef,
    rootPrefixCls,
    showAction,
    mouseLeaveDelay,
    disabled,
    setVisibleDelay,
    inlineElementMousePositionRef
  } = (0, _PopperContext.usePopperContext)();
  const {
    floatingRef,
    arrowRef,
    placement,
    getScrollParents,
    referenceHidden
  } = (0, _usePopper.usePopperPopup)({
    ...popperProps,
    inline: inline ? {
      get x() {
        return inlineElementMousePositionRef.current.x;
      },
      get y() {
        return inlineElementMousePositionRef.current.y;
      }
    } : undefined
  }, {
    skipped: disabled || !visible
  });
  const {
    handlePopupFocus,
    handlePopperMouseDownCapture
  } = (0, _usePopperAction.usePopperPopupAction)({}, {
    referenceElementRef,
    floatingRef,
    visible,
    setVisibleDelay
  });

  // Transition
  const animatedNodeRef = _react.default.useRef(null);

  // FocusLock
  const shards = (0, _reactUtilities.useConstant)(() => [floatingRef, referenceElementRef]);

  // getPopupContainer
  const ctxGetPopupContainer = (0, _reactProvider.usePopupPortalContainer)();
  const portalContainer = _react.default.useMemo(() => {
    if (popupPortalContainer === false) {
      // non portal
      return false;
    }
    if (popupPortalContainer === null) {
      // default portal
      return null;
    }
    return () => {
      if ((0, _util.isFunction)(popupPortalContainer)) {
        return popupPortalContainer(referenceElementRef.current);
      }
      if (popupPortalContainer) {
        return popupPortalContainer.current;
      }
      if (referenceElementRef.current) {
        return (getPopupContainer !== null && getPopupContainer !== void 0 ? getPopupContainer : ctxGetPopupContainer)(referenceElementRef.current);
      }
      return null;
    };
  }, [ctxGetPopupContainer, getPopupContainer, popupPortalContainer, referenceElementRef]);
  const mergedInitialFocusRef = initialFocus === null ? null : initialFocus !== null && initialFocus !== void 0 ? initialFocus : animatedNodeRef;

  // 🐛
  // useLogger('PopperPopup', {
  //   placement,
  //   referenceHidden: drillingCtx?.referenceHidden,
  // });

  // 🐛
  // console.log('Transition', {
  //   status,
  //   placement,
  // });

  const isHidden = referenceHidden !== null && referenceHidden !== void 0 ? referenceHidden : false;
  const isNoEntry = mouseLeaveDelay === 0;
  const lockProps = {
    'data-popper-animated': true,
    style: {
      display: 'flex',
      flexDirection: 'column',
      flex: 'auto',
      minWidth: 0,
      minHeight: 0,
      outline: 'none'
    },
    tabIndex: -1,
    role: 'dialog',
    onMouseEnter: handleMouseEnter,
    onMouseLeave: handleMouseLeave,
    onMouseDownCapture: handlePopperMouseDownCapture,
    onKeyDown: handleKeyDown,
    onClick: e => {
      if (popupClickStopPropagation) {
        e.stopPropagation();
      }
    }
  };
  return /*#__PURE__*/_react.default.createElement(_PopperDrillingContext.PopperDrillingContextProvider, {
    value: {
      getScrollParents,
      referenceHidden
    }
  }, /*#__PURE__*/_react.default.createElement(_PopperPopupContext.PopperPopupContextProvider, {
    value: {
      placement
    }
  }, /*#__PURE__*/_react.default.createElement(_portal.default, {
    container: portalContainer
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(rootPrefixCls, popupClassName),
    style: {
      position: 'absolute',
      display: 'flex',
      flexDirection: 'column',
      ...(isHidden || !(status !== 'exited') ? {
        visibility: 'hidden',
        opacity: 0
      } : {}),
      // 自动隐藏、无法移入以及 visible false（包括了 exit-active 阶段）的时候要让鼠标无法移入
      // TODO 需要补充测试。用 visible false 可以解决点击 popup 中取消按钮 close 过程中误触发 mouseEnter 导致再次 popup 再次打开
      ...(isHidden || isNoEntry || !visible ? {
        pointerEvents: 'none'
      } : {}),
      ...popupStyle
    },
    ref: floatingRef,
    // 事件先挂在外层因为 FocusLock 的 lockProps 会覆盖 onBlur 和 onFocus https://github.com/theKashey/react-focus-lock/issues/189
    onFocus: handlePopupFocus
  }, /*#__PURE__*/_react.default.createElement(UserTransition, {
    nodeRef: animatedNodeRef,
    timeout: timeout,
    appear: true
    // 一定要在 positioning 状态第一次同步过后再做动画（因为首次进来的 placement 可能经过第一次 positioning 以后就变了），对于依赖 placement 的动画可能会不对，比如说 slide up down
    // https://bytedance.feishu.cn/wiki/wikcn5wrtUjqegriRR1zKpyzb3c
    ,
    in: visible && placement !== null,
    mountOnEnter: false,
    unmountOnExit: false,
    extraProps: {
      placement: placement !== null && placement !== void 0 ? placement : 'left' // 这里是随便给了一个 placement，由于我们只会在有 placement 的时候做动画，这个 fallback 只用作类型校验
    }
  }, innerStatus =>
  /*#__PURE__*/
  //  A11y: When Popover opens, focus is sent to PopoverContent. When it closes, focus is returned to the reference. 这里一定要把定位元素和动画元素分开是因为他们的 transform 属性大概率会冲突
  _react.default.createElement(_FocusLock.FocusLock, {
    ref: animatedNodeRef,
    initialFocusRef: mergedInitialFocusRef,
    restoreFocus:
    // hover 的时候关闭 restoreFocus，因为现在没有很好的办法解决 restoreFocus 到 reference 的时候导致 popup 再次打开的问题
    !showAction.includes(_usePopperAction.PopperActionEnum.mouseEnter) && enableRestoreFocus,
    enableFocusLock: enableFocusLock,
    shards: shards
    // 注意这里必须需要和 pointerEvents none 样式同步，当前元素是 pointerEvents none 时，focusLock 会 focus 到 body 或者 reference 导致一些问题。
    // 不能用 animatedVisible 是因为关闭的时候就要立即取消，不能等待动画，不然就太慢了。
    ,
    isDisabled: isHidden || !visible || isNoEntry,
    as: "section",
    className: `${rootPrefixCls}__content`,
    lockProps: innerStatus === 'exited' ? {
      ...lockProps,
      style: {
        ...lockProps.style,
        opacity: 0
      }
    } : lockProps
  }, children, /*#__PURE__*/_react.default.createElement("div", {
    className: `${rootPrefixCls}__arrow`,
    style: {
      position: 'absolute',
      lineHeight: 0 // 不加这个在 safari 上高度会变大
    },
    ref: arrowRef
  }, arrow)))))));
};
exports.PopperPopup = PopperPopup;
