// @universe-design/react/lib/shared/components/popper/usePopper.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.usePopperPopup = usePopperPopup;
exports.usePopperShared = usePopperShared;
var React = _interopRequireWildcard(require("react"));
var _core = require("@floating-ui/core");
var _dom = require("@floating-ui/dom");
var _uniq = _interopRequireDefault(require("lodash/uniq"));
var _rxjs = require("rxjs");
var _operators = require("rxjs/operators");
var _util = require("../../../util");
var _hooks = require("../../../hooks");
var _reactUtilities = require("@universe-design/react-utilities");
var _useImmer = require("use-immer");
var _PopperDrillingContext = require("./PopperDrillingContext");
var _useComplementEvent = require("../../../hooks/useComplementEvent");
var _PopperBoundary = require("./PopperBoundary");
var _PopperContext = require("./PopperContext");
/* eslint-disable max-nested-callbacks */
/* eslint-disable @typescript-eslint/no-use-before-define */

const defaults = {
  middleware: []
};
/**
 * 这个 hook 的逻辑会在 invisible first render 运行，需要尽可能精简
 */
function usePopperShared() {
  /**
   * ================================
   *
   * 从 Popup 注册到全局的方法
   *
   * ================================
   */
  const popperLatestValues = (0, _reactUtilities.useConstant)(() => ({
    updateByReference: undefined,
    update: undefined,
    resetPlacement: undefined
  }));

  /**
   * ================================
   *
   * Reference 相关的 ref 方法
   *
   * ================================
   */
  const [referenceElement, setReferenceElement] = React.useState(null); // 用 state 存 reference element 的原因是 reference 的 element 变化很可能是子元素自己 rerender 导致的，而非父级 rerender。通过在 refCall 中调用 setState 让子元素的 element 变化父级可感知，这样父就可以对子进行事件的重新订阅。

  // 用于获取真实的 element
  const referenceElementRef = React.useRef(null);
  const referenceRef = (0, _reactUtilities.useProxyRef)(null, function (reference) {
    var _popperLatestValues$u;
    // 🐛
    // console.log('PopperReference ref', reference);
    if (reference) {
      // 保证 referenceElementRef 里面一定不会存 null 值，这样 getPopupContainer 才能保证一定取到 reference element
      // **注意** 由于这里存的是上一次 rerender 后的非空值，本次 render 可能会变化造成问题
      referenceElementRef.current = (0, _useComplementEvent.getElementFromRefOrVirtualElement)(reference);
    } else {
      referenceElementRef.current = null;
    }
    setReferenceElement(reference ? referenceElementRef.current : null);
    (_popperLatestValues$u = popperLatestValues.updateByReference) === null || _popperLatestValues$u === void 0 ? void 0 : _popperLatestValues$u.call(popperLatestValues, reference);
  });
  return {
    referenceRef,
    referenceElementRef,
    referenceElement,
    popperLatestValues
  };
}
/**
 * 注意这个 hook 的实现要尽可能 delay，避免在没有渲染的时候进行不必要的计算
 */
function usePopperPopup(props) {
  let {
    skipped = false
  } = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  const ctxBoundary = (0, _PopperBoundary.usePopperBoundaryContext)();
  const {
    placement: placementProp = 'bottom',
    offset: offsetProp = 0,
    eventListeners = {},
    shift: shiftProp = true,
    rootBoundary = 'viewport',
    boundary = ctxBoundary !== null && ctxBoundary !== void 0 ? ctxBoundary : 'clippingAncestors',
    boundaryPadding = 16,
    arrowPadding = 0,
    middlewares: middlewaresProp = defaults.middleware,
    matchReferenceWidth = false,
    flip: flipProp = true,
    enableMaxHeight: maxHeight = true,
    enableAutoHide: autoHide = true,
    onPositioned,
    inline: inlineProp
  } = props;
  const {
    scroll = true,
    resize = true,
    resizeObserver = true
  } = eventListeners;
  const {
    referenceElement,
    referenceRef,
    referenceElementRef,
    popperLatestValues
  } = (0, _PopperContext.usePopperContext)();

  /**
   * Refs
   */

  const onPositionedLatest = (0, _hooks.useEventCallback)(onPositioned);

  /**
   * Scroll & Resize Listener
   */
  const drillingCtx = (0, _PopperDrillingContext.usePopperDrillingContext)();
  const getOwnScrollParents = (0, _hooks.useEventCallback)(() => {
    var _drillingCtx$getScrol;
    const nodes = (_drillingCtx$getScrol = drillingCtx === null || drillingCtx === void 0 ? void 0 : drillingCtx.getScrollParents()) !== null && _drillingCtx$getScrol !== void 0 ? _drillingCtx$getScrol : [];
    const elements = [referenceElementRef.current, floatingRefBeforeProxyRefCallback.current].filter(_reactUtilities.isElement);
    if (!elements.length) {
      return nodes;
    }
    const scrollParents = (0, _uniq.default)(elements.map(el => (0, _dom.getOverflowAncestors)(el)).flat());
    return (0, _uniq.default)([...scrollParents, ...nodes]);
  });

  /**
   * Floating resize
   */
  const floatingResizeObservableDisposer = React.useRef(null);
  const unsubscribeToFloatingResize = () => {
    if (floatingResizeObservableDisposer.current) {
      floatingResizeObservableDisposer.current();
    }
  };
  const subscribeToFloatingResize = () => {
    const floating = floatingRefBeforeProxyRefCallback.current;
    if (skipped || !resizeObserver) {
      return;
    }
    if (floating) {
      floatingResizeObservableDisposer.current = _reactUtilities.resizeObservable.subscribe(floating, updateLatest);
    }
  };

  /**
   * Scoll
   */
  const floatingScrollParentListenersDisposer = React.useRef(null);
  const unsubscribeToScroll = () => {
    if (floatingScrollParentListenersDisposer.current) {
      floatingScrollParentListenersDisposer.current();
    }
  };
  const subscribeToScroll = () => {
    if (skipped || !scroll) {
      return;
    }
    const nodes = getOwnScrollParents();
    if (!nodes.length) {
      return;
    }
    nodes.forEach(node => {
      node.addEventListener('scroll', updateLatest);
    });
    floatingScrollParentListenersDisposer.current = () => {
      nodes.forEach(node => {
        node.removeEventListener('scroll', updateLatest);
      });
    };
  };

  /**
   * Unsubscribe listeners during ref phase
   */
  (0, _reactUtilities.useWillUnmount)(() => {
    unsubscribeToFloatingResize();
    unsubscribeToScroll();
  });

  // 用于给 useProxyRef callback 中调用的函数获取最新的 floating
  const floatingRefBeforeProxyRefCallback = React.useRef(null);
  const floatingRef = (0, _reactUtilities.useProxyRef)(null, function (floating) {
    floatingRefBeforeProxyRefCallback.current = floating;
    // 🐛
    // console.log('PopperPopup ref', floating);
    _updateLatest({
      reference: referenceRef.current,
      floating
    });

    // Floating resize
    unsubscribeToFloatingResize();
    subscribeToFloatingResize();

    // Listen to scroll
    unsubscribeToScroll();
    subscribeToScroll();
  });
  const arrowRef = React.useRef(null);

  /**
   * State
   */
  const [data, updateData] = (0, _useImmer.useImmer)(() => ({
    placement: null,
    referenceHidden: false
  }));
  const middleware = React.useMemo(() => {
    /**
     * overflow options
     */
    const overflowOptions = {
      boundary,
      padding: boundaryPadding,
      rootBoundary
    };

    /**
     * Options
     */
    const flipOptions = (() => {
      if (!flipProp) {
        return {};
      }
      return {
        fallbackStrategy: 'bestFit',
        ...overflowOptions,
        ...(flipProp === true ? {} : flipProp)
      };
    })();
    const shiftOptions = (() => {
      if (!shiftProp) {
        return {};
      }
      return {
        ...overflowOptions,
        crossAxis: true,
        // 默认开启 crossAxis 为了实现当页面中 flip 怎么都放不下的时候盖住 reference
        ...(shiftProp === true ? {} : shiftProp)
      };
    })();
    return [inlineProp ? (0, _core.inline)(inlineProp) : undefined,
    // {
    //   name: 'log',
    //   fn({ elements: { reference, floating }, rects }) {
    //     console.log(
    //       reference === referenceRef.current,
    //       floating === floatingRef.current,
    //       'inline',
    //       JSON.stringify(inlineProp),
    //       'rects.reference',
    //       rects.reference,
    //       'rects.floating',
    //       rects.floating,
    //       'placement',
    //       { placementProp, dataPlacement: data.placement }
    //     );
    //     return {};
    //   },
    // },
    matchReferenceWidth ? {
      name: 'matchReferenceWidth',
      fn(_ref) {
        let {
          placement: _placement,
          rects,
          elements: {
            floating
          }
        } = _ref;
        if (['top', 'bottom'].includes(_placement.split('-')[0])) {
          rects.floating.width = rects.reference.width;
          floating.style.width = `${rects.reference.width}px`;
        }
        return {};
      }
    } : undefined, (0, _core.offset)(offsetProp), autoHide ? (0, _core.hide)({
      ...overflowOptions,
      padding: 0 // hide 逻辑不考虑安全边距，这样可以让 popup 在边缘时保留
    }) : undefined, flipProp ? (0, _core.flip)(flipOptions) : undefined, shiftProp ? (0, _core.shift)(shiftOptions) : undefined, maxHeight && !inlineProp // TODO 暂时在 inline 开启的关闭 maxHeight https://github.com/floating-ui/floating-ui/issues/1570
    ? (0, _core.size)({
      ...overflowOptions,
      apply(_ref2) {
        let {
          availableHeight
        } = _ref2;
        const floatingEl = floatingRef.current;
        if (!floatingEl) {
          return;
        }
        Object.assign(floatingEl.style, {
          maxHeight: `${availableHeight}px`
        });
      }
    }) : undefined, {
      name: 'arrow',
      fn(args) {
        if (arrowRef.current !== null) {
          return (0, _core.arrow)({
            element: arrowRef.current,
            padding: arrowPadding
          }).fn(args);
        }
        return {};
      }
    }, {
      name: 'transformOrigin',
      fn: _ref3 => {
        let {
          placement: _placement,
          rects,
          middlewareData: {
            arrow: _arrow
          }
        } = _ref3;
        if (!_arrow) {
          return {};
        }
        const {
          x: arrowX,
          y: arrowY
        } = _arrow;
        const {
          width: popperW,
          height: popperH
        } = rects.floating;
        let transformOrigin = '';
        if (_placement.startsWith('right')) {
          transformOrigin = `0 ${(arrowY !== null && arrowY !== void 0 ? arrowY : 0) / popperH * 100}%`;
        }
        if (_placement.startsWith('left')) {
          transformOrigin = `100% ${(arrowY !== null && arrowY !== void 0 ? arrowY : 0) / popperH * 100}%`;
        }
        if (_placement.startsWith('top')) {
          transformOrigin = `${(arrowX !== null && arrowX !== void 0 ? arrowX : 0) / popperW * 100}% 100%`;
        }
        if (_placement.startsWith('bottom')) {
          transformOrigin = `${(arrowX !== null && arrowX !== void 0 ? arrowX : 0) / popperW * 100}% 0`;
        }
        return {
          data: {
            transformOrigin
          }
        };
      }
    }];
  }, [arrowPadding, autoHide, boundary, boundaryPadding, flipProp, floatingRef, inlineProp, matchReferenceWidth, maxHeight, offsetProp, rootBoundary, shiftProp]);
  const inputPopperElements$ = (0, _reactUtilities.useConstant)(() => new _rxjs.BehaviorSubject({
    reference: null,
    floating: null,
    placement: placementProp,
    middleware: []
  }));
  const computePosition$ = (0, _reactUtilities.useConstant)(() => inputPopperElements$.pipe((0, _operators.switchMap)(_ref4 => {
    let {
      reference,
      floating,
      placement,
      middleware: _middleware
    } = _ref4;
    if (!reference || !floating) {
      return _rxjs.EMPTY;
    }
    return (0, _rxjs.from)((0, _dom.computePosition)(reference, floating, {
      middleware: _middleware,
      placement,
      strategy: 'absolute'
    }).then(o => ({
      ...o,
      floating
    })));
  })));
  React.useEffect(() => {
    const subscription = computePosition$.subscribe({
      next(computeData) {
        // 这里直接操作 dom 元素的 style 是因为这样可以不用等 react 调度，但是要保证 jsx 那边不要写 style，否则两边同名属性会互相覆盖
        const {
          x,
          y,
          placement,
          middlewareData: {
            hide: {
              referenceHidden = false
            } = {},
            arrow: {
              x: arrowX,
              y: arrowY
            } = {},
            transformOrigin
          },
          floating
        } = computeData;
        // 🐛
        // console.log('beforePositionApplied', computeData);
        updateData(draft => {
          draft.placement = computeData.placement;
          draft.referenceHidden = referenceHidden;
        });
        /**
         * Apply Floating Style
         */
        (function () {
          floating.dataset.popperPlacement = placement;
          floating.style.top = '0px';
          floating.style.left = '0px';
          floating.style.transform = `translate3d(${Math.round(x)}px, ${Math.round(y)}px, 0)`;
        })();
        /**
         * Apply Arrow Style
         */
        (function () {
          const arrowEl = arrowRef.current;
          if (!arrowEl) {
            return;
          }
          // cleanup
          arrowEl.style.top = '';
          arrowEl.style.left = '';
          arrowEl.style.bottom = '';
          arrowEl.style.right = '';
          // eslint-disable-next-line max-nested-callbacks
          const staticSide = (() => {
            switch (placement.split('-')[0]) {
              case 'top':
                return 'bottom';
              case 'right':
                return 'left';
              case 'bottom':
                return 'top';
              case 'left':
                return 'right';
              default:
                throw new Error('Placement main side is not valid.');
            }
          })();
          arrowEl.style.left = (0, _util.isNonNullable)(arrowX) ? `${arrowX}px` : '';
          arrowEl.style.top = (0, _util.isNonNullable)(arrowY) ? `${arrowY}px` : '';
          arrowEl.style[staticSide] = '0px';
        })();
        /**
         * Apply Animated Node Style
         */
        (function () {
          const animatedElement = floating.querySelector('[data-popper-animated]');
          if (!animatedElement) {
            return;
          }
          Object.assign(animatedElement.style, transformOrigin);
        })();
        (0, _util.isFunction)(onPositionedLatest) && onPositionedLatest(computeData);
      },
      error(error) {
        console.error(error);
      }
    });
    return () => {
      subscription.unsubscribe();
    };
    // 这些都是不会变的值，如果有变化的值是不能放进来的，我们不希望多次订阅
  }, [computePosition$, onPositionedLatest, updateData]);
  const _update = React.useCallback(_ref5 => {
    let {
      reference,
      floating
    } = _ref5;
    inputPopperElements$.next({
      reference,
      floating,
      placement: placementProp,
      middleware: [...middleware, ...middlewaresProp].filter(_util.isNonNullable)
    });
  }, [inputPopperElements$, placementProp, middleware, middlewaresProp]);
  const _updateLatest = (0, _hooks.useEventCallback)(_update);
  const update = React.useCallback(() => {
    _update({
      reference: referenceRef.current,
      floating: floatingRef.current
    });
  }, [_update, floatingRef, referenceRef]);
  const updateLatest = (0, _hooks.useEventCallback)(update);

  /**
   * 在 popup 出现的时候往 popper 注入一些方法，这个写法主要是为了首屏尽可能少运行逻辑代码，尽可能把 visible 以后用到的逻辑都收敛到 popup 中
   */
  popperLatestValues.update = updateLatest;
  popperLatestValues.updateByReference = (0, _hooks.useEventCallback)(reference => {
    _update({
      reference,
      floating: floatingRef.current
    });
  });
  popperLatestValues.resetPlacement = React.useCallback(() => {
    updateData(draft => {
      draft.placement = null;
    });
  }, [updateData]);
  (0, _reactUtilities.useWillUnmount)(() => {
    popperLatestValues.resetPlacement = undefined;
    popperLatestValues.update = undefined;
    popperLatestValues.updateByReference = undefined;
  });

  /**
   * Window resize
   */
  React.useEffect(() => {
    if (skipped || !resize) {
      return;
    }
    const win = (0, _util.getWindow)(referenceElement);
    win.addEventListener('resize', updateLatest);
    return () => {
      win.removeEventListener('resize', updateLatest);
    };
  }, [referenceElement, resize, skipped, updateLatest]);

  /**
   * Reference resize
   */
  React.useEffect(() => {
    // 保证在 skipped 变化时能重新订阅
    if (skipped || !resizeObserver) {
      return;
    }
    if (!referenceElement) {
      return;
    }
    return _reactUtilities.resizeObservable.subscribe(referenceElement, updateLatest);
  }, [referenceElement, resizeObserver, skipped, updateLatest]);
  (0, _hooks.useIsomorphicLayoutEffect)(update, [update]);
  return {
    placement: data.placement,
    referenceHidden: (drillingCtx === null || drillingCtx === void 0 ? void 0 : drillingCtx.referenceHidden) || data.referenceHidden,
    floatingRef,
    arrowRef,
    getScrollParents: getOwnScrollParents
  };
}
