// @universe-design/styles/cjs/vc/index.dark.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.vctoken_meeting_btn_icon_schedule = exports.vctoken_meeting_btn_icon_new_meeting = exports.vctoken_meeting_btn_icon_minutes = exports.vctoken_meeting_btn_bg_toolbar = exports.vctoken_meeting_btn_bg_share_screen_sharing = exports.vctoken_meeting_btn_bg_share_screen_s = exports.vctoken_meeting_btn_bg_share_screen_pause = exports.vctoken_meeting_btn_bg_share_screen_m = exports.vctoken_meeting_btn_bg_schedule_s = exports.vctoken_meeting_btn_bg_schedule_m = exports.vctoken_meeting_btn_bg_on_gray = exports.vctoken_meeting_btn_bg_new_meeting_s = exports.vctoken_meeting_btn_bg_new_meeting_m = exports.vctoken_meeting_btn_bg_minutes_s = exports.vctoken_meeting_btn_bg_minutes_m = exports.vctoken_meeting_bg_video_off_auto_layout = exports.vctoken_meeting_bg_video_call = exports.vctoken_meeting_bg_upcoming_hover = exports.vctoken_meeting_bg_upcoming = exports.vctoken_meeting_bg_toolbar = exports.vctoken_meeting_bg_subtitle = exports.vctoken_meeting_bg_side_by_side_view = exports.vctoken_meeting_bg_share_screen_sharing = exports.vctoken_meeting_bg_share_screen_pause = exports.vctoken_meeting_bg_share_panel = exports.vctoken_meeting_bg_share_bar_immersive_mode = exports.vctoken_meeting_bg_push = exports.vctoken_meeting_bg_preview_cam_off = exports.vctoken_meeting_bg_pc_ipad_scroll_tips = exports.vctoken_meeting_bg_ongoing_hover = exports.vctoken_meeting_bg_ongoing = exports.vctoken_meeting_bg_mobile_share_bar_landscape = exports.vctoken_meeting_bg_mobile_share_bar = exports.vctoken_meeting_bg_host_tag = exports.vctoken_meeting_bg_host = exports.vctoken_meeting_bg_float_video = exports.vctoken_meeting_bg_float_transparent = exports.vctoken_meeting_bg_float_pressed = exports.vctoken_meeting_bg_float_mic = exports.vctoken_meeting_bg_float_hover = exports.vctoken_meeting_bg_float = exports.vctoken_meeting_bg_feed = exports.vctoken_meeting_bg_cohost = exports.vctoken_meeting_bg_chat_bubbles_grey_focus = exports.vctoken_meeting_bg_chat_bubbles_grey = exports.vctoken_meeting_bg_chat_bubbles_blue_focus = exports.vctoken_meeting_bg_cam_off = exports.vctoken_meeting_bg_base = exports.vctoken_meeting_bg_action_panel = exports.vctoken_meetin_bg_video_off = void 0;
exports.vctoken_vc_btn_text_selected = exports.vctoken_vc_btn_icon_share_screen = exports.vctoken_vc_btn_icon_selected = exports.vctoken_vc_btn_icon_schedule = exports.vctoken_vc_btn_icon_new_meeting = exports.vctoken_vc_btn_icon_minutes = exports.vctoken_vc_btn_icon_meeting = exports.vctoken_vc_btn_icon_livestream = exports.vctoken_vc_btn_icon_bg_share_screen = exports.vctoken_vc_btn_icon_bg_schedule = exports.vctoken_vc_btn_icon_bg_new_meeting = exports.vctoken_vc_btn_icon_bg_minutes = exports.vctoken_vc_btn_icon_bg_livestream = exports.vctoken_vc_btn_fill_selected = exports.vctoken_vc_btn_bg_share_screen = exports.vctoken_vc_btn_bg_schedule = exports.vctoken_vc_btn_bg_opacity = exports.vctoken_vc_btn_bg_new_meeting = exports.vctoken_vc_btn_bg_minutes = exports.vctoken_vc_btn_bg_livestream = exports.vctoken_vc_btn_bg_float_pressed = exports.vctoken_vc_btn_bg_float_hover = exports.vctoken_vc_btn_bg_float = exports.vctoken_vc_bg_video_call = exports.vctoken_vc_bg_subtitle = exports.vctoken_vc_bg_mobile_share_bar = exports.vctoken_vc_bg_float_video = exports.vctoken_vc_bg_float_2 = exports.vctoken_vc_bg_float = exports.vctoken_vc_bg_card_right = exports.vctoken_vc_bg_card_left = exports.vctoken_vc_bg_card_ended = exports.vctoken_vc_bg_cam_off = exports.vctoken_vc_bg_body = exports.vctoken_vc_annotate_bg_yellow = exports.vctoken_vc_annotate_bg_red = exports.vctoken_vc_annotate_bg_purple = exports.vctoken_vc_annotate_bg_green = exports.vctoken_vc_annotate_bg_blue = exports.vctoken_meeting_text_button = exports.vctoken_meeting_tag_text_cohost = exports.vctoken_meeting_tag_bg_cohost = exports.vctoken_meeting_tag_bg_blue = exports.vctoken_meeting_line_host_tag = exports.vctoken_meeting_line_divider = exports.vctoken_meeting_icon_minutes = exports.vctoken_meeting_icon_disable_on_gray_bg = exports.vctoken_meeting_fill_mask = exports.vctoken_meeting_fill_bar = exports.vctoken_meeting_btn_icon_share_screen = void 0;
exports.vctoken_vc_share_app_border = exports.vctoken_vc_shadow_xs = exports.vctoken_vc_shadow_sm = exports.vctoken_vc_list_icon_bg_upcoming_hover = exports.vctoken_vc_list_icon_bg_upcoming = exports.vctoken_vc_list_icon_bg_ongoing_hover = exports.vctoken_vc_list_icon_bg_ongoing = exports.vctoken_vc_fill_ring = exports.vctoken_vc_fill_hover = void 0;
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
exports.vctoken_meetin_bg_video_off = '#242424';
exports.vctoken_meeting_bg_action_panel = 'rgba(26, 26, 26, 0.98)';
exports.vctoken_meeting_bg_base = '#0a0a0a';
exports.vctoken_meeting_bg_cam_off = '#1f1f1f';
exports.vctoken_meeting_bg_chat_bubbles_blue_focus = '#2d4673';
exports.vctoken_meeting_bg_chat_bubbles_grey = 'rgba(235, 235, 235, 0.08)';
exports.vctoken_meeting_bg_chat_bubbles_grey_focus = 'rgba(235, 235, 235, 0.16)';
exports.vctoken_meeting_bg_cohost = 'rgba(67, 67, 67, 0.9)';
exports.vctoken_meeting_bg_feed = '#373737';
exports.vctoken_meeting_bg_float = '#434343';
exports.vctoken_meeting_bg_float_hover = '#757575';
exports.vctoken_meeting_bg_float_mic = 'rgba(55, 55, 55, 0.8)';
exports.vctoken_meeting_bg_float_pressed = '#5f5f5f';
exports.vctoken_meeting_bg_float_transparent = 'rgba(41, 41, 41, 0.85)';
exports.vctoken_meeting_bg_float_video = '#373737';
exports.vctoken_meeting_bg_host = 'rgba(117, 117, 117, 0.85)';
exports.vctoken_meeting_bg_host_tag = 'rgba(67, 67, 67, 0.9)';
exports.vctoken_meeting_bg_mobile_share_bar = '#121212';
exports.vctoken_meeting_bg_mobile_share_bar_landscape = 'rgba(10, 10, 10, 0.9)';
exports.vctoken_meeting_bg_ongoing = 'rgba(41, 102, 33, 0.6)';
exports.vctoken_meeting_bg_ongoing_hover = 'rgba(65, 158, 52, 0.6)';
exports.vctoken_meeting_bg_pc_ipad_scroll_tips = '#222222';
exports.vctoken_meeting_bg_preview_cam_off = '#282b2e';
exports.vctoken_meeting_bg_push = 'rgba(41, 41, 41, 0.9)';
exports.vctoken_meeting_bg_share_bar_immersive_mode = 'rgba(31, 31, 31, 0.9)';
exports.vctoken_meeting_bg_share_panel = '#101010';
exports.vctoken_meeting_bg_share_screen_pause = '#ffb679';
exports.vctoken_meeting_bg_share_screen_sharing = '#85dc8e';
exports.vctoken_meeting_bg_side_by_side_view = '#141414';
exports.vctoken_meeting_bg_subtitle = 'rgba(235, 235, 235, 0.1)';
exports.vctoken_meeting_bg_toolbar = 'rgba(41, 41, 41, 0.94)';
exports.vctoken_meeting_bg_upcoming = 'rgba(25, 66, 148, 0.6)';
exports.vctoken_meeting_bg_upcoming_hover = 'rgba(38, 85, 182, 0.6)';
exports.vctoken_meeting_bg_video_call = '#1a1a1a';
exports.vctoken_meeting_bg_video_off_auto_layout = '#242424';
exports.vctoken_meeting_btn_bg_minutes_m = 'rgba(42, 45, 105, 0.6)';
exports.vctoken_meeting_btn_bg_minutes_s = '#373d90';
exports.vctoken_meeting_btn_bg_new_meeting_m = 'rgba(25, 66, 148, 0.35)';
exports.vctoken_meeting_btn_bg_new_meeting_s = '#2655b6';
exports.vctoken_meeting_btn_bg_on_gray = '#5f5f5f';
exports.vctoken_meeting_btn_bg_schedule_m = 'rgba(74, 43, 16, 0.6)';
exports.vctoken_meeting_btn_bg_schedule_s = '#8a4a19';
exports.vctoken_meeting_btn_bg_share_screen_m = 'rgba(33, 81, 26, 0.3)';
exports.vctoken_meeting_btn_bg_share_screen_pause = '#3b1a02';
exports.vctoken_meeting_btn_bg_share_screen_s = '#296621';
exports.vctoken_meeting_btn_bg_share_screen_sharing = '#022c07';
exports.vctoken_meeting_btn_bg_toolbar = '#292929';
exports.vctoken_meeting_btn_icon_minutes = '#f8f8f8';
exports.vctoken_meeting_btn_icon_new_meeting = '#f8f8f8';
exports.vctoken_meeting_btn_icon_schedule = '#f8f8f8';
exports.vctoken_meeting_btn_icon_share_screen = '#f8f8f8';
exports.vctoken_meeting_fill_bar = 'rgba(235, 235, 235, 0.15)';
exports.vctoken_meeting_fill_mask = 'rgba(0, 0, 0, 0.6)';
exports.vctoken_meeting_icon_disable_on_gray_bg = 'rgba(232, 232, 232, 0.3)';
exports.vctoken_meeting_icon_minutes = '#1099cc';
exports.vctoken_meeting_line_divider = 'rgba(41, 41, 41, 0.8)';
exports.vctoken_meeting_line_host_tag = 'rgba(235, 235, 235, 0.05)';
exports.vctoken_meeting_tag_bg_blue = 'rgba(46, 101, 209, 0.6)';
exports.vctoken_meeting_tag_bg_cohost = 'rgba(95, 95, 95, 0.8)';
exports.vctoken_meeting_tag_text_cohost = '#e0e0e0';
exports.vctoken_meeting_text_button = '#75a4ff';
exports.vctoken_vc_annotate_bg_blue = '#3370ff';
exports.vctoken_vc_annotate_bg_green = '#34c724';
exports.vctoken_vc_annotate_bg_purple = '#7f3bf5';
exports.vctoken_vc_annotate_bg_red = '#f54a45';
exports.vctoken_vc_annotate_bg_yellow = '#ffc60a';
exports.vctoken_vc_bg_body = '#0a0a0a';
exports.vctoken_vc_bg_cam_off = '#1a1a1a';
exports.vctoken_vc_bg_card_ended = '#5f5f5f';
exports.vctoken_vc_bg_card_left = '#2e374d';
exports.vctoken_vc_bg_card_right = '#536080';
exports.vctoken_vc_bg_float = '#292929';
exports.vctoken_vc_bg_float_2 = '#292929';
exports.vctoken_vc_bg_float_video = '#292929';
exports.vctoken_vc_bg_mobile_share_bar = '#0a0a0a';
exports.vctoken_vc_bg_subtitle = 'rgba(26, 26, 26, 0.6)';
exports.vctoken_vc_bg_video_call = '#1a1a1a';
exports.vctoken_vc_btn_bg_float = '#505050';
exports.vctoken_vc_btn_bg_float_hover = '#757575';
exports.vctoken_vc_btn_bg_float_pressed = '#5f5f5f';
exports.vctoken_vc_btn_bg_livestream = 'rgba(29, 78, 71, 0.8)';
exports.vctoken_vc_btn_bg_minutes = 'rgba(22, 67, 89, 0.8)';
exports.vctoken_vc_btn_bg_new_meeting = '#194294';
exports.vctoken_vc_btn_bg_opacity = 'rgba(80, 80, 80, 0.9)';
exports.vctoken_vc_btn_bg_schedule = 'rgba(104, 58, 18, 0.7)';
exports.vctoken_vc_btn_bg_share_screen = 'rgba(33, 81, 26, 0.8)';
exports.vctoken_vc_btn_fill_selected = 'rgba(235, 235, 235, 0.12)';
exports.vctoken_vc_btn_icon_bg_livestream = '#25665e';
exports.vctoken_vc_btn_icon_bg_minutes = '#135a78';
exports.vctoken_vc_btn_icon_bg_new_meeting = '#3370eb';
exports.vctoken_vc_btn_icon_bg_schedule = 'rgba(138, 74, 25, 0.8)';
exports.vctoken_vc_btn_icon_bg_share_screen = 'rgba(41, 102, 33, 0.8)';
exports.vctoken_vc_btn_icon_livestream = '#f8f8f8';
exports.vctoken_vc_btn_icon_meeting = '#75a4ff';
exports.vctoken_vc_btn_icon_minutes = '#f8f8f8';
exports.vctoken_vc_btn_icon_new_meeting = '#f8f8f8';
exports.vctoken_vc_btn_icon_schedule = '#f8f8f8';
exports.vctoken_vc_btn_icon_selected = '#e8e8e8';
exports.vctoken_vc_btn_icon_share_screen = '#f8f8f8';
exports.vctoken_vc_btn_text_selected = '#ebebeb';
exports.vctoken_vc_fill_hover = 'rgba(235, 235, 235, 0.12)';
exports.vctoken_vc_fill_ring = '#ffffff';
exports.vctoken_vc_list_icon_bg_ongoing = 'rgba(33, 81, 26, 0.6)';
exports.vctoken_vc_list_icon_bg_ongoing_hover = 'rgba(41, 102, 33, 0.6)';
exports.vctoken_vc_list_icon_bg_upcoming = 'rgba(25, 66, 148, 0.6)';
exports.vctoken_vc_list_icon_bg_upcoming_hover = 'rgba(38, 85, 182, 0.6)';
exports.vctoken_vc_shadow_sm = 'rgba(0, 0, 0, 0.5)';
exports.vctoken_vc_shadow_xs = 'rgba(0, 0, 0, 0.05)';
exports.vctoken_vc_share_app_border = 'rgba(235, 235, 235, 0.2)';
