// @universe-design/react/src/tooltip/stories/homeless.stories.tsx
/** @jsxImportSource @emotion/react */
import type { ComponentProps } from 'react';
import React, { useState } from 'react';
import { Global } from '@emotion/react';
import type { TooltipPlacement } from '..';
import { Tooltip } from '..';
import type { DefaultSizeType } from '../..';
import {
  Button,
  Slider,
  Radio,
  Space,
  Checkbox,
  Popover,
  Popconfirm,
  Dropdown,
  Menu,
  Switch,
  IconButton,
  Input,
  InputNumber,
  RangePicker,
  DatePicker,
  ConfigProvider,
  Form,
} from '../..';
import './index.less';
import { default as PlacementDemo } from '../demo/placement';
import AddButtonOutlined from '@universe-design/icons-react/AddButtonOutlined';

export default {
  title: 'Components/Tooltip/临时示例',
  component: Tooltip,
};

export const BasicUsage = Object.assign((props: ComponentProps<typeof Tooltip>) => (
  <Tooltip {...props}>Some text</Tooltip>
));

export const placements = () => (
  <div style={{ marginTop: 200, marginLeft: 200 }}>
    <PlacementDemo />
  </div>
);

const des =
  '超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案';
const largeDes =
  '超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案超长文案';
const smallDes = '一段文案，一段文案';

export const PlacementCustomer = () => {
  const [marginTop, setMarginTop] = useState(100);
  const [marginLeft, setMarginLeft] = useState(100);
  const [placement, setPlacement] = useState<TooltipPlacement>('top');
  const [strSize, setStrSize] = useState<DefaultSizeType>('small');

  return (
    <div style={{ height: '200vh' }}>
      <Space size="large" direction="vertical" style={{ width: '100%' }}>
        <Slider step={100} min={0} max={1000} value={marginTop} onChange={setMarginTop} />
        <Slider step={100} min={0} max={1000} value={marginLeft} onChange={setMarginLeft} />
        <Space align="center">
          <span>方向:</span>
          <Radio.Group
            value={placement}
            onChange={v => {
              setPlacement(v as TooltipPlacement);
              console.log('placementState change', v);
            }}
          >
            <Radio.Button value="top">上</Radio.Button>
            <Radio.Button value="bottom">下</Radio.Button>
            <Radio.Button value="left">左</Radio.Button>
            <Radio.Button value="right">右</Radio.Button>
            <Radio.Button value="topLeft">上左</Radio.Button>
            <Radio.Button value="topRight">上右</Radio.Button>
            <Radio.Button value="bottomLeft">下左</Radio.Button>
            <Radio.Button value="bottomRight">下右</Radio.Button>
            <Radio.Button value="leftTop">左上</Radio.Button>
            <Radio.Button value="leftBottom">左下</Radio.Button>
            <Radio.Button value="rightTop">右上</Radio.Button>
            <Radio.Button value="rightBottom">右下</Radio.Button>
          </Radio.Group>
        </Space>
        <Space align="center">
          <span>内容大小:</span>
          <Radio.Group value={strSize} onChange={v => setStrSize(v as DefaultSizeType)}>
            <Radio.Button value="small">短内容</Radio.Button>
            <Radio.Button value="large">长内容</Radio.Button>
          </Radio.Group>
        </Space>
      </Space>
      <Tooltip placement={placement} title={strSize === 'small' ? smallDes : largeDes}>
        <Button style={{ marginTop, marginLeft }}>超长文案超长文案超长文案</Button>
      </Tooltip>
    </div>
  );
};

export const PlacementCase = () => (
  <div className="ud__tooltip_w">
    <div>
      <Tooltip title={des}>
        <div className="ud__pte" />
      </Tooltip>
      <Tooltip title={des}>
        <div className="ud__pte ud__pte-tr" />
      </Tooltip>
      <Tooltip title={des}>
        <div className="ud__pte ud__pte-br" />
      </Tooltip>
      <Tooltip title={des}>
        <div className="ud__pte ud__pte-bl" />
      </Tooltip>
    </div>
  </div>
);

export const PlacementCase2 = () => (
  <div className="ud__tooltip_w">
    <div>
      <Tooltip title={des} placement="bottom">
        <div className="ud__pte ud__pte-tc" />
      </Tooltip>
      <Tooltip title={des} placement="right">
        <div className="ud__pte ud__pte-lc" />
      </Tooltip>
      <Tooltip title={des} placement="left">
        <div className="ud__pte ud__pte-rc" />
      </Tooltip>
      <Tooltip title={des} placement="top">
        <div className="ud__pte ud__pte-bc" />
      </Tooltip>
    </div>
  </div>
);

export const Disabled = () => (
  <Tooltip placement="left" title={largeDes} disabled>
    <Button>超长文案超长文案超长文案</Button>
  </Tooltip>
);

export const WithTooltip = () => (
  <Tooltip title="title msg demo">
    <Checkbox />
  </Tooltip>
);

export const BodyBorderBug = () => {
  const [formValues, setFormValues] = React.useState<{
    clipParent: 'body' | 'blueBox';
    bodyBorder: boolean;
  }>({
    clipParent: 'body',
    bodyBorder: true,
  });
  return (
    <ConfigProvider getPopupContainer={node => node?.parentNode as HTMLElement}>
      <Global
        styles={{
          body: {
            ...(formValues.bodyBorder ? { border: '1px solid red' } : {}),
            height: '100vh',
          },
        }}
      />
      <p>
        这个例子是为了说明在 body border 存在情况下 dom-align 计算 popup 位置时会超出body 1px。此时如果不给子元素 width:
        max-content 还会导致循环计算错误，导致 popup 的宽度一直缩小
      </p>
      <Form initialValues={formValues} onValuesChange={(_, all) => setFormValues(all)}>
        <Form.Item name="bodyBorder" label="body 是否有 border" valuePropName="checked">
          <Switch />
        </Form.Item>
        <Form.Item name="clipParent" label="切换 clip parent">
          <Radio.Group>
            <Radio value="body">body</Radio>
            <Radio value="blueBox">蓝色框</Radio>
          </Radio.Group>
        </Form.Item>
      </Form>
      <div
        css={{
          border: '1px solid blue',
          float: 'right',
          width: 600,
          height: 600,
          display: 'flex',
          flexDirection: 'column',
          alignItems: 'flex-end',
          ...(formValues.clipParent === 'blueBox'
            ? {
                position: 'relative',
                overflow: 'hidden',
              }
            : {}),
        }}
      >
        <div>
          <Tooltip placement="bottom" title={largeDes}>
            <Button>Tooltip</Button>
          </Tooltip>
        </div>
        <div>
          <Popover placement="bottom" title={largeDes}>
            <Button>Popover</Button>
          </Popover>
        </div>
        <div>
          <Popconfirm placement="bottom" title={largeDes}>
            <Button>Popconfirm</Button>
          </Popconfirm>
        </div>
        <div>
          <Dropdown
            placement="bottomCenter"
            overlay={
              <Menu textOverflow={false}>
                <Menu.Item key="a">{largeDes}</Menu.Item>
                <Menu.Item key="b">{largeDes}</Menu.Item>
                <Menu.Item key="c">{largeDes}</Menu.Item>
              </Menu>
            }
          >
            <Button>Dropdown</Button>
          </Dropdown>
        </div>
      </div>
    </ConfigProvider>
  );
};

export const DisabledButton = () => {
  type DisableProp = { disabled?: boolean };
  const renderTooltip = (Component: React.FunctionComponent<DisableProp>) => (
    <Tooltip title="xxxxx" wrapDisabledChildren>
      <Component disabled />
    </Tooltip>
  );
  return (
    <Space direction="vertical">
      <h5>已经不需要使用 wrapDisabledChildren 的表单元素：</h5>
      <Tooltip title="xxxxx">
        <Checkbox disabled />
      </Tooltip>
      <Tooltip title="xxxxx">
        <Switch disabled />
      </Tooltip>

      <h5>还需要使用 wrapDisabledChildren 的表单元素：</h5>

      <Tooltip title="xxxxx" wrapDisabledChildren>
        <IconButton icon={<AddButtonOutlined />} />
      </Tooltip>
      <Tooltip title="xxxxx" wrapDisabledChildren>
        <Button>BUTTON</Button>
      </Tooltip>
      {[Button, Switch].map(renderTooltip)}
      <Tooltip title="xxxxx" wrapDisabledChildren>
        <Input disabled value="3333" />
      </Tooltip>
      <Tooltip title="xxxxx" wrapDisabledChildren>
        <Input.TextArea disabled value="3333" />
      </Tooltip>
      <Tooltip title="xxxxx" wrapDisabledChildren>
        <DatePicker picker="month" disabled />
      </Tooltip>
      <Tooltip title="xxxxx" wrapDisabledChildren>
        <RangePicker picker="month" />
      </Tooltip>

      <Tooltip title="xxxxx">
        <InputNumber disabled value={3333} />
      </Tooltip>
      <Tooltip title="xxxxx" wrapDisabledChildren>
        <Button type="link" disabled href="baidu.com">
          44
        </Button>
      </Tooltip>
      <Tooltip title="xxxxx" wrapDisabledChildren>
        <Button disabled={true}>Hello world!</Button>
      </Tooltip>
      <Tooltip title="xxxxx" wrapDisabledChildren>
        <select disabled={true}>
          <option value="ssss" />
        </select>
      </Tooltip>
      <Tooltip title="xxxxx" wrapDisabledChildren>
        <input disabled={true} />
      </Tooltip>
      <Tooltip title="xxxxx" wrapDisabledChildren>
        <textarea disabled={true} />
      </Tooltip>
      <Tooltip title="xxxxx" wrapDisabledChildren>
        <IconButton icon={<AddButtonOutlined />} disabled />
      </Tooltip>
    </Space>
  );
};

export const IssueXXX = () => {
  const [vis, setVis] = React.useState(false);
  return (
    <Popover content="tttt" visible={vis}>
      <Tooltip title="">
        <Button
          onClick={() => {
            setVis(_vis => !_vis);
          }}
        >
          123
        </Button>
      </Tooltip>
    </Popover>
  );
};

export const multiLineOverflow = () => {
  const checkOverflow = (ele: Element) => {
    console.log('called: ', ele);
    return ele.scrollHeight > ele.clientHeight;
  };
  return (
    <Tooltip title="overflowed" checkOverflow={checkOverflow}>
      <div
        style={{
          margin: 100,
          width: 200,
          WebkitLineClamp: 2,
          overflow: 'hidden',
          textOverflow: 'ellipsis',
          display: '-webkit-box',
          WebkitBoxOrient: 'vertical',
        }}
      >
        Tooltip will show on mouse enter. Tooltip will show on mouse enter. Tooltip will show on mouse enter. Tooltip
        will show on mouse enter.
      </div>
    </Tooltip>
  );
};
