// @universe-design/react/src/tooltip/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C6Inline from '../demo/6.inline';
import CBasic from '../demo/basic';
import CControlled from '../demo/controlled';
import CDisabledChildren from '../demo/disabledChildren';
import COverflow from '../demo/overflow';
import CPlacement from '../demo/placement';
if (!(C6Inline as Story).storyName) {
  (C6Inline as Story).storyName = '6. 包裹 inline 元素';
  (C6Inline as Story).parameters = {
    docs: {
      description: {
        story:
          '仅 `PopperSidecar={FloatingUIPopperSidecar}` 模式下可用。此模式下，我们会使用 `getClientRects` 获取内联元素的多段矩形区域，并按最合适的方式进行弹出。\n\n',
      },
    },
  };
}
if (!(CBasic as Story).storyName) {
  (CBasic as Story).storyName = '0. 基础用法';
  (CBasic as Story).parameters = {
    docs: {
      description: {
        story: '简单的用法。\n\n',
      },
    },
  };
}
if (!(CControlled as Story).storyName) {
  (CControlled as Story).storyName = '2. 受控模式';
  (CControlled as Story).parameters = {
    docs: {
      description: {
        story: '通过`visible`和`onVisibleChange`可以按需控制文字提示的渲染。\n\n',
      },
    },
  };
}
if (!(CDisabledChildren as Story).storyName) {
  (CDisabledChildren as Story).storyName = '5. 用于禁用态表单组件';
  (CDisabledChildren as Story).parameters = {
    docs: {
      description: {
        story:
          'Tooltip 依赖子元素触发的 onMouseEnter、onMouseLeave 等事件。\n但在 Chrome 等浏览器上，禁用态的表单元素，可能并不能触发这些事件，相关 Bug 讨论： [chromium](https://bugs.chromium.org/p/chromium/issues/detail?id=120132)。\n\n可以使用`wrapDisabledChildren`属性让 UD 自动为 tooltip 子组件包裹一个`span`标签，解决 tooltip 无法触发的问题，建议仅用在禁用态表单元素上。\n\n',
      },
    },
  };
}
if (!(COverflow as Story).storyName) {
  (COverflow as Story).storyName = '3. 溢出检测';
  (COverflow as Story).parameters = {
    docs: {
      description: {
        story:
          '设置`checkOverflow={true}`，则只有在内容溢出时才触发文字提示显示, 也可以传入函数来自己做是否溢出的逻辑判断。\n\n',
      },
    },
  };
}
if (!(CPlacement as Story).storyName) {
  (CPlacement as Story).storyName = '1. 位置';
  (CPlacement as Story).parameters = {
    docs: {
      description: {
        story: '位置有 12 个方向。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Tooltip/官网示例',
  tags: ['autodocs'],
};

export { C6Inline, CBasic, CControlled, CDisabledChildren, COverflow, CPlacement };
