// @universe-design/react-utilities/lib-commonjs/hooks/composeRefs.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.composeRefs = composeRefs;
exports.supportRef = supportRef;
exports.updateRef = updateRef;
exports.useMergedRefs = useMergedRefs;
var React = _interopRequireWildcard(require("react"));
var _reactIs = require("react-is");
/**
 * 可选 ref
 *
 * @public
 */

/**
 * 非空 ref
 *
 * @public
 */

const composedRefCache = new WeakMap();

/**
 * 更新 ref 的值
 *
 * @public
 */
function updateRef(ref, instance) {
  if (typeof ref === 'function') {
    ref(instance);
  } else {
    // eslint-disable-next-line @typescript-eslint/no-extra-semi
    ref.current = instance;
  }
}
function composeTwoRefs(ref1, ref2) {
  if (ref1 && ref2) {
    const ref1Cache = composedRefCache.get(ref1) || new WeakMap();
    composedRefCache.set(ref1, ref1Cache);
    const composedRef = ref1Cache.get(ref2) || (instance => {
      updateRef(ref1, instance);
      updateRef(ref2, instance);
    });
    ref1Cache.set(ref2, composedRef);
    return composedRef;
  }
  if (!ref1) {
    return ref2;
  } else {
    return ref1;
  }
}
function isClassComponent(component) {
  var _component$prototype;
  return typeof component === 'function' && Boolean((_component$prototype = component.prototype) === null || _component$prototype === void 0 ? void 0 : _component$prototype.isReactComponent);
}
function isFunctionComponent(component) {
  var _component$prototype2;
  return typeof component === 'function' && !((_component$prototype2 = component.prototype) !== null && _component$prototype2 !== void 0 && _component$prototype2.isReactComponent);
}

/**
 * 判断是否支持 ref
 *
 * @public
 */
function supportRef(node) {
  if (! /*#__PURE__*/React.isValidElement(node)) {
    return false;
  }
  let _node = node;
  while ((0, _reactIs.isMemo)(_node)) {
    // eslint-disable-next-line @typescript-eslint/no-extra-semi
    ({
      type: _node
    } = _node);
  }
  const {
    type
  } = _node;
  if (isClassComponent(type) || isFunctionComponent(type)) {
    return false;
  }
  return true;
}

/**
 * 组合多个 ref
 *
 * @public
 */
function composeRefs() {
  for (var _len = arguments.length, refs = new Array(_len), _key = 0; _key < _len; _key++) {
    refs[_key] = arguments[_key];
  }
  if (refs.length === 2) {
    // micro-optimize the hot path
    return composeTwoRefs(refs[0], refs[1]) || null;
  }
  const composedRef = refs.slice(1).reduce((semiCombinedRef, refToInclude) => composeTwoRefs(semiCombinedRef, refToInclude), refs[0]);
  return composedRef || null;
}

/**
 * 合并多个 ref。注意这里传入的 refs 会作为 useMemo 的 deps，数量不能随 render 变化。
 *
 * @public
 * @param refs - 多个 refs
 * @returns 合并后的 ref
 */
function useMergedRefs() {
  for (var _len2 = arguments.length, refs = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
    refs[_key2] = arguments[_key2];
  }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  return React.useMemo(() => composeRefs(...refs), [...refs]);
}
