// @universe-design/react-utilities/lib-commonjs/hooks/useCloneElementRef.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useCloneElementRef = exports.FindDOMNode = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var React = _interopRequireWildcard(require("react"));
var ReactDOM = _interopRequireWildcard(require("react-dom"));
var _util = require("../util");
var _dom = require("../dom");
var _composeRefs = require("./composeRefs");
/* eslint-disable react/no-find-dom-node */
/* eslint-disable react/prefer-stateless-function */

const getChildrenRef = child => child === null || child === void 0 ? void 0 : child.ref;

/**
 * 用于获取子组件的 dom ref
 *
 * @public
 */
class FindDOMNode extends React.Component {
  constructor() {
    super(...arguments);
    (0, _defineProperty2.default)(this, "node", void 0);
    (0, _defineProperty2.default)(this, "hasRefFuncBeenCalled", false);
    (0, _defineProperty2.default)(this, "isRefSupported", true);
    (0, _defineProperty2.default)(this, "updateChildRef", node => {
      if (this.node === node) {
        return;
      }
      if (this.domOnly && node !== null && !(0, _dom.isHTMLElement)(node) && !(0, _dom.isSVGElement)(node)) {
        return;
      }
      (0, _composeRefs.updateRef)(this.props.childRef, node);
      this.node = node;
    });
    (0, _defineProperty2.default)(this, "updateRefWhenLifecycle", () => {
      // 如果不是 domOnly，那就不需要额外操作 ref 了
      if (!this.domOnly) {
        return;
      }

      // 如果是 domOnly，并且不支持 ref 那就需要在 lifecycle 做更新
      const {
        children
      } = this.props;
      if (!(0, _composeRefs.supportRef)(children)) {
        this.updateChildRef(ReactDOM.findDOMNode(this));
        return;
      }

      // 如果是 domOnly，并且支持 ref，需要兜底一种特殊情况：组件支持 ref 但是内部没有实际调用 ref
      if (!this.hasRefFuncBeenCalled) {
        this.updateChildRef(ReactDOM.findDOMNode(this));
        this.hasRefFuncBeenCalled = false;
      }
    });
    (0, _defineProperty2.default)(this, "updateRefWhenSupportRef", node => {
      const {
        children
      } = this.props;
      const origRef = getChildrenRef(children);
      if (origRef) {
        (0, _composeRefs.updateRef)(origRef, node);
      }
      if (node === null) {
        this.updateChildRef(null);
        this.hasRefFuncBeenCalled = false;
        return;
      }
      this.hasRefFuncBeenCalled = true;
      if (this.domOnly && node && !(0, _dom.isElement)(node)) {
        this.updateChildRef(ReactDOM.findDOMNode(this));
        return;
      }
      this.updateChildRef(node);
    });
  }
  get domOnly() {
    const {
      domOnly = true
    } = this.props;
    return domOnly;
  }
  componentDidMount() {
    this.updateRefWhenLifecycle();
  }
  componentDidUpdate() {
    this.updateRefWhenLifecycle();
  }
  componentWillUnmount() {
    if (!this.isRefSupported) {
      (0, _composeRefs.updateRef)(this.props.childRef, null);
      this.hasRefFuncBeenCalled = false;
    }
  }
  render() {
    this.isRefSupported = false;
    const {
      children: childrenProp,
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      childRef,
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      domOnly,
      ...restProps
    } = this.props;
    if (! /*#__PURE__*/React.isValidElement(childrenProp)) {
      return childrenProp;
    }
    // forward props
    const children = /*#__PURE__*/React.cloneElement(childrenProp, restProps);
    if (!(0, _composeRefs.supportRef)(children)) {
      return children;
    }
    this.isRefSupported = true;
    return /*#__PURE__*/React.cloneElement(children, {
      ref: this.updateRefWhenSupportRef
    });
  }
}

/**
 * 给 child node clone ref
 *
 * @public
 */
exports.FindDOMNode = FindDOMNode;
const useCloneElementRef = function (_ref) {
  let {
    children,
    ref
  } = _ref;
  let {
    skipped = false,
    domOnly
  } = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  if (skipped || !(0, _util.isNonNullable)(children)) {
    return children;
  }
  return /*#__PURE__*/React.createElement(FindDOMNode, {
    domOnly: domOnly,
    childRef: ref
  }, children);
};
exports.useCloneElementRef = useCloneElementRef;
