// @universe-design/react-utilities/lib-commonjs/hooks/useResizeObserver.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useResizeObserver = exports.getResizeObserver = exports.createResizeObserver = void 0;
var _react = require("react");
var _reactDom = require("react-dom");
var _warning = require("../warning");
var _util = require("../util");
var _dom = require("../dom");
var _resizeObservable = require("../dom/resizeObservable");
var _useIsomorphicLayoutEffect = require("./useIsomorphicLayoutEffect");
var _useLatest = require("./useLatest");
var _useProxyRef = require("./useProxyRef");
/* eslint-disable no-undef */

// window.onerror = console.trace;

/**
 * Create a resize observer with subscribe and unsubscribe methods
 *
 * @internal
 */
const createResizeObserver = () => {
  const callbacks = new Set();
  return {
    observer: new _resizeObservable.resizeObservable.ResizeObserver((entries, observer) => {
      // entries.forEach(e =>
      //   console.log(
      //     '%cResizeObserver.callback',
      //     'background: red;',
      //     e.target.className,
      //     e.target.textContent,
      //     e.borderBoxSize[0]
      //   )
      // );

      // 使用 flushSync 同步执行状态更新避免在 resize 时造成的闪屏问题
      (0, _reactDom.flushSync)(() => {
        (0, _reactDom.unstable_batchedUpdates)(() => {
          for (const callback of callbacks) {
            callback(entries, observer);
          }
        });
      });
    }),
    subscribe: callback => callbacks.add(callback),
    unsubscribe: callback => callbacks.delete(callback)
  };
};
exports.createResizeObserver = createResizeObserver;
let _resizeObserver;

/**
 * @internal
 */
const getResizeObserver = () => !_resizeObserver ? _resizeObserver = createResizeObserver() : _resizeObserver;

/**
 * A React hook that fires a callback whenever ResizeObserver detects a change to its size
 * https://github.com/jaredLunde/react-hook/tree/master/packages/resize-observer#readme
 *
 * @public
 * @param callback - Invoked with a single `ResizeObserverEntry` any time
 *   the `target` resizes
 */
exports.getResizeObserver = getResizeObserver;
const useResizeObserver = function (callback) {
  let {
    disabled = false
  } = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  const resizeObserver = getResizeObserver();
  const storedCallback = (0, _useLatest.useLatest)(callback);
  const observationStateRef = (0, _react.useRef)({
    refCalledTimes: 0,
    observingNode: null
  });
  const targetRef = (0, _useProxyRef.useProxyRef)(null, node => {
    if (disabled) {
      return;
    }
    const refCalledTimes = observationStateRef.current.refCalledTimes + 1;
    observationStateRef.current.refCalledTimes = refCalledTimes;
    Promise.resolve().then(() => {
      if (refCalledTimes !== observationStateRef.current.refCalledTimes) {
        // 如果在一个 task 中调用多次，只保留最后一次
        return;
      }
      const {
        observingNode
      } = observationStateRef.current;
      if (observingNode === node) {
        // 如果需要观察的节点和之前的节点相同，则跳过
        return;
      }
      observingNode && resizeObserver.observer.unobserve(observingNode);
      if (!(0, _dom.isElement)(node)) {
        // 如果没有需要观察的节点
        observationStateRef.current.observingNode = null;
        return;
      }
      process.env.NODE_ENV !== 'production' && (0, _warning.warn)(!(node && (0, _util.isNonReplacedElement)(node)), 'useResizeObserver', "'Observations' do not fire for non-replaced inline Elements, refer to https://www.w3.org/TR/resize-observer/#intro'", node);
      resizeObserver.observer.observe(node, {
        box: 'border-box'
      });
      observationStateRef.current.observingNode = node;
      // console.log(
      //   '%cresizeObserver.observer.observe',
      //   'background: blue;',
      //   node.className,
      //   node.textContent
      // );
    });
  });
  (0, _useIsomorphicLayoutEffect.useIsomorphicLayoutEffect)(() => {
    if (disabled) {
      return;
    }
    let didUnsubscribe = false;
    const _callback = (entries, observer) => {
      if (didUnsubscribe || targetRef.current && (0, _util.isNonReplacedElement)(targetRef.current)) {
        return;
      }

      // eslint-disable-next-line @typescript-eslint/prefer-for-of
      for (let i = 0; i < entries.length; i++) {
        const entry = entries[i];
        if (entry.target === targetRef.current) {
          storedCallback.current(entry, observer);
        }
      }
    };
    resizeObserver.subscribe(_callback);
    return () => {
      didUnsubscribe = true;
      resizeObserver.unsubscribe(_callback);
    };
  }, [disabled, resizeObserver, storedCallback, targetRef]);
  return targetRef;
};

/**
 * The callback type of useResizeObserver
 *
 * @public
 */
exports.useResizeObserver = useResizeObserver;
