// @universe-design/react-utilities/lib-commonjs/hooks/useStyleSheetOnce.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StyleSheetOnce = void 0;
exports.useStyleSheetOnce = useStyleSheetOnce;
var _react = _interopRequireDefault(require("react"));
var _sheet = require("@emotion/sheet");
var _hash = _interopRequireDefault(require("@emotion/hash"));
var _useIsomorphicLayoutEffect = require("./useIsomorphicLayoutEffect");
var _reactProvider = require("@universe-design/react-provider");
/**
 * 全局锁和 styleSheet 实例
 */
const augmentedCache = {};

/**
 * 需求
 * 1. emotion global 那种插入一段内容到全局，可以重复加入
 * 2. 通过一个样式的 hash 作为 key，如果已经 inserted 后续的就不加入了
 */

/**
 * @public
 * useStyleSheetOnce
 * 全局插入一次 css 样式，适用于静态样式场景
 */
function useStyleSheetOnce(styles) {
  let {
    disabled = false,
    atLeastOne = false
  } = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  // 计算 serialized style，主要是 hash styles 作为 name
  const serialized = _react.default.useMemo(() => !disabled ? {
    name: (0, _hash.default)(styles),
    styles
  } : {
    name: '',
    styles: ''
  }, [styles, disabled]);
  const cache = (0, _reactProvider.useEmotionCache)();
  const initialize = _react.default.useCallback(() => {
    if (disabled) {
      return;
    }
    const key = `${cache.key}-global`;
    // 创建专用 sheet 和 augmentedCache value
    if (!augmentedCache[serialized.name]) {
      const sheet = new _sheet.StyleSheet({
        key,
        nonce: cache.sheet.nonce,
        container: cache.sheet.container,
        speedy: cache.sheet.isSpeedy
      });
      augmentedCache[serialized.name] = {
        injectedCount: 0,
        sheet
      };
      if (cache.sheet.tags.length) {
        [sheet.before] = cache.sheet.tags;
      }
      // shouldCache true false 都无所谓，因为我们自己做了额外的全局 cache 逻辑
      cache.insert('', serialized, sheet, false);
    }
  }, [cache, disabled, serialized]);

  // 尽可能早初始化
  initialize();
  (0, _useIsomorphicLayoutEffect.useIsomorphicLayoutEffect)(() => {
    if (disabled) {
      return;
    }
    // render 比其他的 cleanup 函数先调用，所以当真正执行到这里的时候 cacheValue 已经被清理了，需要再做一遍初始化
    initialize();
    const cacheValue = augmentedCache[serialized.name];
    const {
      sheet
    } = cacheValue;
    cacheValue.injectedCount += 1;
    return () => {
      cacheValue.injectedCount -= 1;
      if (cacheValue.injectedCount === 0 && !atLeastOne) {
        sheet.flush();
        delete augmentedCache[serialized.name];
      }
    };
  }, [atLeastOne, disabled, initialize, serialized.name]);
}
/**
 * @public
 */
const StyleSheetOnce = _ref => {
  let {
    styles
  } = _ref;
  useStyleSheetOnce(styles);
  return null;
};
exports.StyleSheetOnce = StyleSheetOnce;
