// @universe-design/react/lib/shared/components/picker/Comp/panels/time-panel/Column.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _rcVirtualList = _interopRequireDefault(require("@universe-design/rc-virtual-list"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _clsx = _interopRequireDefault(require("clsx"));
var _configProvider = require("../../../../../../config-provider");
var _react = _interopRequireWildcard(require("react"));
var _hooks = require("../../../../../../hooks");
var _util = require("../../../../../../util");
var _list = require("../../../../list");
var _List = require("../../../../list/List");
var _Context = require("../../../Context");
var _utils = require("../../../utils");
var _date = require("../../../utils/date");
var _stylex2 = require("../../../../../../stylex");
/* eslint-disable @typescript-eslint/no-magic-numbers */

function Column(props) {
  const {
    units,
    onPanelChange,
    value,
    hideDisabledOptions,
    mode,
    generateConfig,
    ulWidth
  } = props;
  const inDatePicker = (0, _Context.usePickerContextSelector)(o => o.inDatePicker);
  // 是否被嵌套在 datePicker 中
  const cssScope = (0, _configProvider.useCssScope)('time-panel');
  const prefixCls = cssScope();
  const itemCount = inDatePicker ? 264 / 32 : 7.5;
  const listRef = (0, _react.useRef)(null);
  const vListRef = (0, _react.useRef)(null);
  const liRefs = (0, _react.useRef)(new Map());
  const scrollToItem = _react.default.useCallback(function (_value) {
    let duration = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 120;
    const li = liRefs.current.get(_value);
    const list = listRef.current;
    if (li && list) {
      // 第一个元素 margin-top: 1;
      (0, _utils.scrollTo)(list, li.offsetTop - 1, duration);
    }
    const vList = vListRef.current;
    if (vList) {
      vList.scrollTo({
        key: _value,
        align: 'top'
      });
    }
  }, []);
  (0, _hooks.useEffectOnce)(() => {
    // TODO 这里不用异步会获取不到 li 的尺寸和位置，后面需要看一下
    requestAnimationFrame(() => {
      scrollToItem(Number(value), 0);
    });
  });
  (0, _hooks.useUpdateEffect)(_react.default.useCallback(() => {
    scrollToItem(Number(value));
  }, [scrollToItem, value]));
  const isSameTimeDateType = (time1, time2) => {
    if (mode === 'timeCombine') {
      return time1 !== -1 ? (0, _date.isSameTime)(generateConfig, time1, time2) : false;
    } else {
      return false;
    }
  };
  const isSame = (time1, time2) => {
    if (mode === 'timeCombine') {
      return isSameTimeDateType(time1, time2);
    }
    return time1 === time2;
  };
  const renderItem = unit => {
    if (hideDisabledOptions && unit.disabled) {
      return null;
    }
    return /*#__PURE__*/_react.default.createElement(_list.ListItem, {
      key: Number(unit.value),
      ref: element => {
        liRefs.current.set(Number(unit.value), element);
      },
      onClick: () => {
        if (unit.disabled) {
          return;
        }
        onPanelChange && onPanelChange(unit.value);
      },
      disabled: unit.disabled,
      selected: isSame(value, unit.value),
      showCheckmark: mode === 'timeCombine'
    }, /*#__PURE__*/_react.default.createElement(_list.ListItemTitle, null, unit.label));
  };
  return (() => mode === 'timeCombine' ? /*#__PURE__*/_react.default.createElement(_rcVirtualList.default, {
    data: units !== null && units !== void 0 ? units : [],
    itemKey: unit => Number(unit.value),
    fullHeight: false,
    itemHeight: 32,
    height: 32 * itemCount,
    virtual: !_util.__TEST__,
    ref: vListRef,
    style: {
      width: '100%'
    }
  }, unit => {
    const item = renderItem(unit);
    if (item) {
      return /*#__PURE__*/_react.default.createElement("div", null, item);
    }
  }) : /*#__PURE__*/_react.default.createElement(_List.ListBox, {
    className: `${prefixCls}__column`
  }, inDatePicker && /*#__PURE__*/_react.default.createElement("span", {
    className: (0, _clsx.default)(`${prefixCls}__column-title`, (0, _stylex.default)(_stylex2.cssTypographyTitle4))
  }, props.title), /*#__PURE__*/_react.default.createElement(_List.ListScrollView, {
    style: {
      width: ulWidth,
      height: itemCount * 32,
      position: 'relative' // ele.scrollTop is offset of closest relative positioned parent.
    },
    className: `${prefixCls}__column-scroll-list`,
    ref: listRef
  }, (units !== null && units !== void 0 ? units : []).map(renderItem), /*#__PURE__*/_react.default.createElement("div", {
    style: {
      paddingBottom: (itemCount - 1) * 32
    }
  }))))();
}
var _default = exports.default = Column;
