// @universe-design/react/lib/shared/components/picker/Comp/panels/time-panel/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = TimePanel;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _utils = require("../../../utils");
var _NormalTimePanel = _interopRequireDefault(require("./NormalTimePanel"));
var _CombineTimePanel = _interopRequireDefault(require("./CombineTimePanel"));
var _date = require("../../../utils/date");
const countBoolean = boolList => boolList.filter(bool => bool !== false).length;
function TimePanel(props) {
  const {
    generateConfig,
    prefixCls,
    operationRef,
    value,
    format,
    onPanelChange,
    mode = 'time',
    showSecond: showSecondProp,
    showMinute: showMinuteProp,
    showHour: showHourProp,
    use12Hours: use12HoursProp
  } = props;
  const showSecond = showSecondProp !== null && showSecondProp !== void 0 ? showSecondProp : format ? _date.generateFormat.showSecond(format) : false;
  const showMinute = showMinuteProp !== null && showMinuteProp !== void 0 ? showMinuteProp : format ? _date.generateFormat.showMinute(format) : true;
  const showHour = showHourProp !== null && showHourProp !== void 0 ? showHourProp : format ? _date.generateFormat.showHour(format) : true;
  const use12Hours = use12HoursProp !== null && use12HoursProp !== void 0 ? use12HoursProp : format ? _date.generateFormat.is12Hour(format) : false;
  const bodyOperationRef = _react.default.useRef();

  // ======================= Keyboard =======================
  const [activeColumnIndex, setActiveColumnIndex] = _react.default.useState(-1);
  const columnsCount = countBoolean([showSecond, use12Hours]);
  operationRef && (operationRef.current = {
    onKeyDown: event => (0, _utils.createKeyDownHandler)(event, {
      onLeftRight: diff => {
        setActiveColumnIndex((activeColumnIndex + diff + columnsCount) % columnsCount);
      },
      onUpDown: diff => {
        if (activeColumnIndex === -1) {
          setActiveColumnIndex(0);
        } else if (bodyOperationRef.current) {
          bodyOperationRef.current.onUpDown(diff);
        }
      },
      onEnter: () => {
        onPanelChange(value || generateConfig.getNow());
        setActiveColumnIndex(-1);
      }
    }),
    onBlur: () => {
      setActiveColumnIndex(-1);
    }
  });
  return /*#__PURE__*/_react.default.createElement("div", null, (() => {
    if (mode === 'timeCombine') {
      return /*#__PURE__*/_react.default.createElement(_CombineTimePanel.default, (0, _extends2.default)({}, props, {
        use12Hours: use12Hours
      }));
    } else if (mode === 'time') {
      return /*#__PURE__*/_react.default.createElement("div", {
        className: `${prefixCls}-time-panel`
      }, /*#__PURE__*/_react.default.createElement(_NormalTimePanel.default, (0, _extends2.default)({}, props, {
        showMinute: showMinute,
        showHour: showHour,
        use12Hours: use12Hours,
        showSecond: showSecond,
        activeColumnIndex: activeColumnIndex
      })));
    }
  })());
}
