// @universe-design/react/lib/shared/components/picker/Comp/panels/time-panel/NormalTimePanel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _configProvider = require("../../../../../../config-provider");
var _Context = require("../../../../../../date-picker/Context");
var _clsx = _interopRequireDefault(require("clsx"));
var _react = _interopRequireDefault(require("react"));
var _List = require("../../../../list/List");
var _Context2 = require("../../../Context");
var _utils = require("../../../utils");
var _Column = _interopRequireDefault(require("./Column"));
require("./style");
/* eslint-disable @typescript-eslint/no-magic-numbers */

function shouldUnitsUpdate(prevUnits, nextUnits) {
  if (prevUnits.length !== nextUnits.length) {
    return true;
  }
  // if any unit's disabled status is different, the units should be re-evaluted
  for (let i = 0; i < prevUnits.length; i += 1) {
    if (prevUnits[i].disabled !== nextUnits[i].disabled) {
      return true;
    }
  }
  return false;
}
function generateUnits(start, end, step, disabledUnits) {
  const units = [];
  for (let i = start; i <= end; i += step) {
    units.push({
      label: (0, _utils.leftPad)(i, 2),
      value: i,
      disabled: (disabledUnits || []).includes(i)
    });
  }
  return units;
}
function NormalTimePanel(props) {
  var _ref, _props$value;
  const {
    generateConfig,
    prefixCls,
    activeColumnIndex,
    showSecond = false,
    showMinute = true,
    showHour = true,
    use12Hours,
    hourStep = 1,
    minuteStep = 1,
    secondStep = 1,
    disabledHours,
    disabledMinutes,
    disabledSeconds,
    hideDisabledOptions,
    onPanelChange,
    locale,
    popupStyle
  } = props;
  const {
    value: valueFromCtx
  } = (0, _Context2.usePickerContext)();
  const datePickerContext = (0, _Context.useDatePickerContext)();
  const cssScope = (0, _configProvider.useCssScope)('time-panel');
  const classNames = (0, _clsx.default)(datePickerContext.showTime && cssScope('--in-datepicker'));
  const value = (_ref = (_props$value = props.value) !== null && _props$value !== void 0 ? _props$value : valueFromCtx) !== null && _ref !== void 0 ? _ref : -1;
  const parseWidth = popupStyle && parseFloat(popupStyle === null || popupStyle === void 0 ? void 0 : popupStyle.width);
  const ulWidth = parseWidth ? Math.floor(parseWidth / (() => {
    return [showSecond, showHour, showMinute, use12Hours].filter(item => item).length;
  })()) : 70;
  const columns = [];
  let isPM = false;
  const originHour = value === -1 ? -1 : generateConfig.getHour(value);
  let hour = originHour;
  const minute = value === -1 ? -1 : generateConfig.getMinute(value);
  const second = value === -1 ? -1 : generateConfig.getSecond(value);
  const setTime = (isNewPM, newHour, newMinute, newSecond) => {
    let newDate = value === -1 ? generateConfig.getNow() : value;
    const mergedHour = Math.max(0, newHour);
    const mergedMinute = Math.max(0, newMinute);
    const mergedSecond = Math.max(0, newSecond);
    newDate = (0, _utils.setTime)(generateConfig, newDate, !use12Hours || !isNewPM ? mergedHour : mergedHour + 12, mergedMinute, mergedSecond);
    return newDate;
  };

  // ========================= Unit =========================
  const rawHours = generateUnits(0, 23, hourStep, disabledHours && disabledHours());
  const memorizedRawHours = (0, _utils.useMemo)(() => rawHours,
  // eslint-disable-next-line react-hooks/exhaustive-deps
  rawHours, shouldUnitsUpdate);

  // Should additional logic to handle 12 hours
  if (use12Hours && hour) {
    isPM = hour >= 12; // -1 means should display AM
    hour %= 12;
  }
  const hours = _react.default.useMemo(() => {
    if (!use12Hours) {
      return memorizedRawHours || [];
    }
    if (memorizedRawHours) {
      return memorizedRawHours.filter(isPM ? hourMeta => hourMeta.value >= 12 : hourMeta => hourMeta.value < 12).map(hourMeta => {
        const hourValue = hourMeta.value % 12;
        const hourLabel = hourValue === 0 ? '12' : (0, _utils.leftPad)(hourValue, 2);
        return {
          ...hourMeta,
          label: hourLabel,
          value: hourValue
        };
      }) || [];
    }
    return [];
  }, [use12Hours, isPM, memorizedRawHours]);
  const minutes = generateUnits(0, 59, minuteStep, disabledMinutes && disabledMinutes(originHour));
  const seconds = generateUnits(0, 59, secondStep, disabledSeconds && disabledSeconds(originHour, minute));

  // ======================== Render =======================
  function addColumnNode(node, columnValue, units, onColumnSelect) {
    columns.push({
      node: /*#__PURE__*/_react.default.cloneElement(node, {
        prefixCls: `${prefixCls}-time-panel`,
        value: columnValue,
        active: activeColumnIndex === columns.length,
        onPanelChange: onColumnSelect,
        units,
        hideDisabledOptions
      }),
      onPanelChange: onColumnSelect,
      value: columnValue,
      units
    });
  }
  use12Hours && addColumnNode(/*#__PURE__*/_react.default.createElement(_Column.default, {
    key: "12hours",
    value: value,
    ulWidth: ulWidth,
    generateConfig: generateConfig
  }),
  // 0是上午1是下午
  isPM ? 1 : 0, [{
    label: locale.morning,
    value: 0,
    disabled: false
  }, {
    label: locale.afternoon,
    value: 1,
    disabled: false
  }], num => {
    onPanelChange(setTime(Boolean(num), hour, minute, second));
  });
  showHour && addColumnNode(/*#__PURE__*/_react.default.createElement(_Column.default, {
    key: "hour",
    ulWidth: ulWidth,
    value: value,
    generateConfig: generateConfig,
    title: locale.hour
  }), hour, hours, num => {
    onPanelChange(setTime(isPM, num, minute, second));
  });
  showMinute && addColumnNode(/*#__PURE__*/_react.default.createElement(_Column.default, {
    key: "minute",
    value: value,
    ulWidth: ulWidth,
    generateConfig: generateConfig,
    title: locale.minute
  }), minute, minutes, num => {
    onPanelChange(setTime(isPM, hour, num, second));
  });
  showSecond && addColumnNode(/*#__PURE__*/_react.default.createElement(_Column.default, {
    key: "second",
    ulWidth: ulWidth,
    value: value,
    generateConfig: generateConfig,
    title: locale.second
  }), second, seconds, num => {
    onPanelChange(setTime(isPM, hour, minute, num));
  });
  return /*#__PURE__*/_react.default.createElement(_List.ListBox, {
    cascaded: true,
    className: classNames
  }, columns.map(_ref2 => {
    let {
      node
    } = _ref2;
    return node;
  }));
}
var _default = exports.default = NormalTimePanel;
