// @universe-design/react-ai/lib-esm/shared/components/LottieAnimation.js
import _extends from "@babel/runtime/helpers/esm/extends";
import * as React from 'react';
import lottie from 'lottie-web';
import { updateRef, useMergedRefs, useEffectOnce } from '@universe-design/react-utilities';
export default /*#__PURE__*/React.forwardRef((props, ref) => {
  const {
    animationData,
    onSegmentPlayEnd,
    lottieAnimationRef: lottieAnimationRefProp,
    ...htmlProps
  } = props;
  const containerRef = React.useRef(null);
  const mergedContainerRef = useMergedRefs(ref, containerRef);
  const lottieAnimationRef = React.useRef();
  const mergedLottieAnimationRef = useMergedRefs(lottieAnimationRefProp, lottieAnimationRef);
  useEffectOnce(() => {
    if (containerRef.current) {
      const animation = lottie.loadAnimation({
        animationData,
        container: containerRef.current,
        // the dom element that will contain the animation
        renderer: 'svg',
        loop: false,
        autoplay: false
      });
      mergedLottieAnimationRef && updateRef(mergedLottieAnimationRef, animation);
      return () => {
        animation.destroy();
        mergedLottieAnimationRef && updateRef(mergedLottieAnimationRef, null);
      };
    }
  });
  React.useEffect(() => {
    const animation = lottieAnimationRef.current;
    if (!animation || !onSegmentPlayEnd) {
      return;
    }
    animation.addEventListener('complete', onSegmentPlayEnd);
    return () =>
    // animation 已经提前销毁，lottie 内部 removeEventListener 的时候没有判空
    animation._cbs && animation.removeEventListener('complete', onSegmentPlayEnd);
  }, [onSegmentPlayEnd]);
  return /*#__PURE__*/React.createElement("div", _extends({
    ref: mergedContainerRef
  }, htmlProps));
});
