// @universe-design/react-ai/lib-esm/shared/components/ThinkingAnimationBase.js
import _extends from "@babel/runtime/helpers/esm/extends";
import * as React from 'react';
import LottieAnimation from './LottieAnimation';
import { useEventCallback } from '@universe-design/react-utilities';
const initial = 'initial';
export const starting = 'starting';
export const looping = 'looping';
export const ending = 'ending';
export const ThinkingAnimationBase = /*#__PURE__*/React.forwardRef((props, ref) => {
  const {
    thinking,
    segments,
    ...htmlProps
  } = props;
  const animationRef = React.useRef(null);
  const statusRef = React.useRef(initial);
  const onSegmentPlayEnd = useEventCallback(() => {
    if (statusRef.current === ending && thinking) {
      var _animationRef$current;
      (_animationRef$current = animationRef.current) === null || _animationRef$current === void 0 ? void 0 : _animationRef$current.playSegments(segments[starting], false);
      statusRef.current = starting;
      return;
    }
    if (statusRef.current === ending && !thinking) {
      statusRef.current = initial;
      return;
    }
    if (statusRef.current === starting && thinking) {
      var _animationRef$current2;
      (_animationRef$current2 = animationRef.current) === null || _animationRef$current2 === void 0 ? void 0 : _animationRef$current2.playSegments(segments[looping], false);
      statusRef.current = looping;
      return;
    }
    if (statusRef.current === starting && !thinking) {
      var _animationRef$current3;
      (_animationRef$current3 = animationRef.current) === null || _animationRef$current3 === void 0 ? void 0 : _animationRef$current3.playSegments(segments[ending], false);
      statusRef.current = ending;
      return;
    }
    if (statusRef.current === looping && thinking) {
      var _animationRef$current4;
      (_animationRef$current4 = animationRef.current) === null || _animationRef$current4 === void 0 ? void 0 : _animationRef$current4.playSegments(segments[looping], false);
      return;
    }
    if (statusRef.current === looping && !thinking) {
      var _animationRef$current5;
      (_animationRef$current5 = animationRef.current) === null || _animationRef$current5 === void 0 ? void 0 : _animationRef$current5.playSegments(segments[ending], false);
      statusRef.current = ending;
    }
  });
  React.useEffect(() => {
    // start play
    if (thinking && statusRef.current === initial) {
      var _animationRef$current6;
      (_animationRef$current6 = animationRef.current) === null || _animationRef$current6 === void 0 ? void 0 : _animationRef$current6.playSegments(segments[starting], true);
      statusRef.current = starting;
    }
  }, [segments, thinking]);
  return /*#__PURE__*/React.createElement(LottieAnimation, _extends({
    ref: ref,
    lottieAnimationRef: animationRef,
    onSegmentPlayEnd: onSegmentPlayEnd
  }, htmlProps));
});
