// @universe-design/react/es/shared/components/highlighter/index.js
import React from 'react';
import { findAll } from 'highlight-words-core';
import { createContext, useContext } from 'use-context-selector';
const HightLightContext = createContext({
  disabled: false
});
const defaultTextRenderer = text => /*#__PURE__*/React.createElement("span", null, text);
const defaultMarkRenderer = text => /*#__PURE__*/React.createElement("mark", null, text);
export const Highlighter = function (_ref) {
  let {
    autoEscape = true,
    caseSensitive = false,
    findChunks,
    sanitize,
    searchWords,
    textToHighlight,
    markRenderer = defaultMarkRenderer,
    textRenderer = defaultTextRenderer
  } = _ref;
  const chunks = findAll({
    autoEscape,
    caseSensitive,
    findChunks,
    sanitize,
    searchWords,
    textToHighlight
  });
  return /*#__PURE__*/React.createElement("span", null, chunks.map((chunk, index) => {
    const text = textToHighlight.substr(chunk.start, chunk.end - chunk.start);
    if (chunk.highlight) {
      const markNode = markRenderer(text);
      return /*#__PURE__*/React.isValidElement(markNode) ? /*#__PURE__*/React.cloneElement(markNode, {
        key: index
      }) : /*#__PURE__*/React.cloneElement(defaultMarkRenderer(text), {
        key: index
      });
    }
    const textNode = textRenderer(text);
    return /*#__PURE__*/React.isValidElement(textNode) ? /*#__PURE__*/React.cloneElement(textNode, {
      key: index
    }) : /*#__PURE__*/React.cloneElement(defaultTextRenderer(text), {
      key: index
    });
  }));
};
export const HighlightProvider = HightLightContext.Provider;
if (process.env.NODE_ENV !== "production") {
  HighlightProvider.displayName = 'HighlightProvider';
}
export const HighlightConsumer = _ref2 => {
  let {
    markRenderer,
    children
  } = _ref2;
  const {
    searchWords,
    disabled
  } = useContext(HightLightContext);
  if (typeof children === 'string' && searchWords && searchWords.length && !disabled) {
    return /*#__PURE__*/React.createElement(Highlighter, {
      textToHighlight: children,
      searchWords: searchWords,
      markRenderer: markRenderer
    });
  }
  return /*#__PURE__*/React.createElement(React.Fragment, null, children);
};
if (process.env.NODE_ENV !== "production") {
  HighlightConsumer.displayName = 'HighlightConsumer';
}
