// @byted-service/bytedtrace/lib/configsClient/staticConfigurator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StaticConfigurator = void 0;
const tslib_1 = require("tslib");
const configProvider = tslib_1.__importStar(require("../confProvider"));
class StaticConfigurator {
    async pullGlobalTracerConfig() {
        return configProvider.DEFAULT_GLOBAL_TRACER_CONFIG;
    }
    async pullBreakerConfig() {
        return configProvider.DEFAULT_CB_CONFIG;
    }
    async pullSamplerConfig() {
        const config = Object.assign({}, configProvider.DEFAULT_SAMPLER_CONFIG);
        const customSamplingRateProbability = process.env.BYTED_TRACE_CUSTOM_SAMPLING_RATE_PROBABILITY;
        const customSamplingMode = process.env.BYTED_TRACE_CUSTOM_SAMPLING_MODE;
        const customSamplingCountPerMinAdaptive = process.env.BYTED_TRACE_CUSTOM_SAMPLING_COUNT_PER_MIN_ADAPTIVE;
        const defaultStrategy = Object.assign({}, config.defaultStrategy);
        if (customSamplingRateProbability) {
            defaultStrategy.samplingRateProbability =
                parseFloat(customSamplingRateProbability) || 0.001;
        }
        if (customSamplingMode) {
            defaultStrategy.samplingMode = customSamplingMode;
        }
        if (customSamplingCountPerMinAdaptive) {
            defaultStrategy.samplingCountPerMinAdaptive =
                parseInt(customSamplingCountPerMinAdaptive, 10) || 15;
        }
        config.defaultStrategy = defaultStrategy;
        return config;
    }
    register() {
        // do nothing
    }
    async syncInit() {
        // do nothing
    }
    asyncInit() {
        // do nothing
    }
    close() {
        // do nothing
    }
}
exports.StaticConfigurator = StaticConfigurator;
