// @universe-design/react-utilities/lib-commonjs/domain-events/dom.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DomainEventTarget = exports.DomainEvent = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _requestIdleCallback = require("../dom/requestIdleCallback");
// Domains can only emit their own events, but they can subscribe to their sub-domains events.

/**
 * This is an abstract class of a domain event.
 *
 * @alpha
 */
class DomainEvent {
  constructor(_ref) {
    let {
      detail,
      bubbles = true,
      cancelable = true
    } = _ref;
    (0, _defineProperty2.default)(this, "type", void 0);
    (0, _defineProperty2.default)(this, "detail", void 0);
    (0, _defineProperty2.default)(this, "timeStamp", Date.now());
    // eventId = '';
    (0, _defineProperty2.default)(this, "_cancelable", void 0);
    (0, _defineProperty2.default)(this, "_bubbles", void 0);
    (0, _defineProperty2.default)(this, "_defaultPrevented", false);
    (0, _defineProperty2.default)(this, "_propagationStopped", false);
    (0, _defineProperty2.default)(this, "preventDefault", () => {
      if (!this._cancelable) {
        return;
      }
      this._defaultPrevented = true;
    });
    (0, _defineProperty2.default)(this, "isDefaultPrevented", () => this._defaultPrevented);
    (0, _defineProperty2.default)(this, "stopPropagation", () => {
      this._propagationStopped = true;
    });
    (0, _defineProperty2.default)(this, "isPropagationStopped", () => !this._bubbles || this._propagationStopped);
    this._cancelable = cancelable;
    this._bubbles = bubbles;
    this.detail = detail;
  }
}

/**
 * Domain event handler callback function type
 *
 * @alpha
 */
exports.DomainEvent = DomainEvent;
/**
 * `DomainEventTarget` class is similar with DOM EventTarget API which contains a callbacks map and some functions to add listeners and invoke listeners.
 *
 * @alpha
 */
class DomainEventTarget {
  constructor() {
    (0, _defineProperty2.default)(this, "_callbacksMap", new WeakMap());
    (0, _defineProperty2.default)(this, "parentEventTarget", null);
    (0, _defineProperty2.default)(this, "addEventListener", (type, cb) => {
      const _cbSet = this._callbacksMap.get(type);
      const cbSet = _cbSet !== null && _cbSet !== void 0 ? _cbSet : new Set();
      if (!_cbSet) {
        this._callbacksMap.set(type, cbSet);
      }
      cbSet.add(cb);
      return () => {
        cbSet.delete(cb);
      };
    });
    (0, _defineProperty2.default)(this, "invokeEventListeners", event => {
      var _this$_callbacksMap$g;
      const cbSet = (_this$_callbacksMap$g = this._callbacksMap.get(event.constructor)) !== null && _this$_callbacksMap$g !== void 0 ? _this$_callbacksMap$g : new Set();
      for (const cb of cbSet) {
        try {
          cb(event);
        } catch (err) {
          // The logic below might not be correct, mainly trying to simulate error handling in DOM events.
          (0, _requestIdleCallback.requestIdleCallback)(() => {
            throw err;
          });
        }
      }
    });
    (0, _defineProperty2.default)(this, "dispatchEvent", event => {
      let currentEventTarget = this;
      while (currentEventTarget) {
        currentEventTarget.invokeEventListeners(event);
        if (event.isPropagationStopped()) {
          break;
        }
        currentEventTarget = currentEventTarget.parentEventTarget;
      }
      return event.isDefaultPrevented();
    });
  }
}
exports.DomainEventTarget = DomainEventTarget;
