// @universe-design/stylex/docs/.config/config.mjs
import { defineConfig } from '@universe-design/heft-rig/shared/site-config.mjs'
import { Module } from 'module'
import path from 'path'

import { fileURLToPath } from 'url'

globalThis.__filename = fileURLToPath(import.meta.url)
globalThis.__dirname = path.dirname(__filename)
globalThis.require = Module.createRequire(import.meta.url) 
const { PUBLIC_PATH, ROUTE_PREFIX } = process.env

export default defineConfig({
    mode: 'app',
    routePath: ROUTE_PREFIX ?? '/develop/stylex',
    menus: [
      {
        title: 'Stylex 官网',
        children: [
          {
            title: '介绍',
            path: '/overview',
            altText: '',
            docPath: '../basic/introduce.md',
            isComponent: false,
          }, 
          {
            title: '语法',
            path: '/api',
            altText: '',
            docPath: '../basic/api-doc.md',
            isComponent: false,
          },
          {
            title: '最佳实践',
            path: '/practices',
            altText: '',
            docPath: '../basic/practices.md',
            isComponent: false,
          },
          {
            title: '限制',
            path: '/errors',
            altText: '',
            docPath: '../basic/errors.md',
            isComponent: false,
          },
          // {
          //   title: '设计原理',
          //   path: '/principle',
          //   altText: '',
          //   docPath: '../basic/principle.md',
          //   isComponent: false,
          // },
        ]
      },
    ],
    webpack: (config) => {
      const oneOfRule = config.module?.rules?.[0];
      const rule = oneOfRule.oneOf.find(
        (r) => r.test?.toString() === '/\\.ts$/'
      )
      console.log(rule)

      config.resolve = {
        ...config.resolve,
        alias: {
          '@universe-design/stylex': path.resolve(
            require.resolve('../../lib-commonjs')
          ),
        }
      }
      if (config.output && PUBLIC_PATH) {
        config.output.publicPath = PUBLIC_PATH;
      }
      return config;
    }
})
