// @universe-design/react/es/tabs/ScrollableTabsHeader.js
//.ulvyvdh{flex-wrap:nowrap;}
import u125g0x8 from "./../.stylex/u125g0x8";
//.u1jhi6b8{flex-grow:0;}
import ug24l6d from "./../.stylex/ug24l6d";
//.u1oy3dpc{overflow-x:auto;}
import u1qr8jow from "./../.stylex/u1qr8jow";
//.uo9lwiw{scrollbar-width:none;}
import u9vy1v2 from "./../.stylex/u9vy1v2";
//.um5o6js ::-webkit-scrollbar{height:0px;}
import u1no85ao from "./../.stylex/u1no85ao";
//.uqc6z8f:after{content:"";}
import u1xegxa0 from "./../.stylex/u1xegxa0";
//.u11ef69:after{position:absolute;}
import u5qwpn8 from "./../.stylex/u5qwpn8";
//.uqevjk0:after{pointer-events:none;}
import u1l27p0d from "./../.stylex/u1l27p0d";
//.u19zsqug:after{z-index:2;}
import ujik7se from "./../.stylex/ujik7se";
//.u1oarbwq:after{right:0px;}
import u1yoqygm from "./../.stylex/u1yoqygm";
//.u1yuyem3:after{top:0px;}
import u1anwibm from "./../.stylex/u1anwibm";
//.u1soztmy:after{bottom:6px;}
import u1zgj7v from "./../.stylex/u1zgj7v";
//.us0klr3:after{width:var(--scroll-mask-width);}
import u1w1hd4l from "./../.stylex/u1w1hd4l";
//.u2df6js:after{display:block;}
import u1aoskr8 from "./../.stylex/u1aoskr8";
//.u16igepq:after{background:@udtoken-component-bg-right-mask;}
import ubm18w5 from "./../.stylex/ubm18w5";
//.uuzzvh5:before{content:"";}
import u1klsf9c from "./../.stylex/u1klsf9c";
//.uxf9f1y:before{position:absolute;}
import uua3xia from "./../.stylex/uua3xia";
//.urcipv9:before{pointer-events:none;}
import u16zaoyg from "./../.stylex/u16zaoyg";
//.u13tv88k:before{z-index:2;}
import u19z4mkn from "./../.stylex/u19z4mkn";
//.u8g0ba1:before{left:0px;}
import u1pid3wf from "./../.stylex/u1pid3wf";
//.urdto5n:before{top:0px;}
import ucgpi4a from "./../.stylex/ucgpi4a";
//.u14i0ki9:before{bottom:6px;}
import upyoa5l from "./../.stylex/upyoa5l";
//.ukapjos:before{width:var(--scroll-mask-width);}
import u102c4m8 from "./../.stylex/u102c4m8";
//.uxv2tmp:before{display:block;}
import u1mp4r5t from "./../.stylex/u1mp4r5t";
//.u1801gqn:before{background:@udtoken-component-bg-left-mask;}
import uc4om0q from "./../.stylex/uc4om0q";
//.u10tiqix{min-width:0px;}
import urmav4p from "./../.stylex/urmav4p";
//.uqerorx{flex-grow:1;}
import u99gbyh from "./../.stylex/u99gbyh";
//.u1neuvcm{flex-shrink:1;}
import uldrd5o from "./../.stylex/uldrd5o";
//.uly5x3f{width:100%;}
import u13d3rh8 from "./../.stylex/u13d3rh8";
//.udrzuqr{cursor:not-allowed;}
import u1vjduc2 from "./../.stylex/u1vjduc2";
//.u1qs8jb8{color:@icon-disabled;}
import u14cxjev from "./../.stylex/u14cxjev";
//.u22iagw{display:flex;}
import ukchsq7 from "./../.stylex/ukchsq7";
//.u1ub7u0d{align-self:stretch;}
import ug4hjch from "./../.stylex/ug4hjch";
//.u122n59{align-items:center;}
import ubyhozq from "./../.stylex/ubyhozq";
//.u64fuq3{width:20px;}
import uevmt6d from "./../.stylex/uevmt6d";
//.u1ugzwwg{font-size:12px;}
import u283mfx from "./../.stylex/u283mfx";
//.u1k6fduh{cursor:pointer;}
import u12o6wmt from "./../.stylex/u12o6wmt";
//.u1bu1o99{color:@icon-n1;}
import u1u0vkcg from "./../.stylex/u1u0vkcg";
//.u10pi13n{position:relative;}
import u5jxv71 from "./../.stylex/u5jxv71";
//.u17gl07{background:@udtoken-component-bg-right-mask;}
import u16ye362 from "./../.stylex/u16ye362";
/* eslint-disable react/no-unknown-property */
/** @jsxRuntime classic */
/** @jsx jsx */
/** @jsxFrag React.Fragment */
import stylex from '@universe-design/stylex';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx } from '@universe-design/stylex/react';
import React from 'react';
import { Tab } from './Tab';
import { TabProvider } from './TabContext';
import { useTabsContext } from './TabsContext';
import LeftOutlined from '@universe-design/icons-react/LeftOutlined';
import RightOutlined from '@universe-design/icons-react/RightOutlined';
import { TabsInkBar } from './TabInk';
import { composeRefs, useResizeObserver } from '@universe-design/react-utilities';
import { forwardHTMLRef } from '../util/forwardRef';
import { IconButton } from '../button';
import { resolveSizingBehavior } from './sizing_behavior';
const defaultScrollOption = {
  // 外层有滚动容器的话，外层不跟随滚动
  block: 'nearest',
  behavior: 'smooth'
};
export const ScrollableTabsHeader = forwardHTMLRef((_ref, ref) => {
  let {
    inkOffsetY
  } = _ref;
  const {
    rootCssScope,
    panes,
    scrollButtonGutter,
    tabElementMapRef,
    type,
    rootTabsSizingBehavior
  } = useTabsContext();
  const [scrollButtonVisible, setScrollButtonVisible] = React.useState(false);
  const [scrollButtonDisabled, setScrollButtonDisabled] = React.useState([true, false]);
  const scrollMaskWidth = type === 'line' ? 22 : 0;
  const resizeHandler = () => {
    if (wrapperRef.current) {
      const {
        scrollWidth,
        clientWidth
      } = wrapperRef.current;
      setScrollButtonVisible(scrollWidth - clientWidth > 1);
      checkScrollButtonStatus();
    }
  };
  // 容器尺寸变化（往往是受到挤压或者缩放）
  const wrapperRef = useResizeObserver(resizeHandler);
  // 列表尺寸变化（往往是数据变化）
  const innerListRef = useResizeObserver(resizeHandler);
  const composedRef = composeRefs(wrapperRef, ref);

  /**
   * 当 tabBar 滚动时，调整 ScrollButton 的禁用态
   */
  const checkScrollButtonStatus = React.useCallback(() => {
    if (!wrapperRef.current) {
      return;
    }
    const leftButtonDisabled = wrapperRef.current.scrollLeft === 0;
    const rightButtonDisabled = Math.round(wrapperRef.current.scrollLeft + wrapperRef.current.clientWidth) >= Math.round(wrapperRef.current.scrollWidth);
    setScrollButtonDisabled([leftButtonDisabled, rightButtonDisabled]);
  }, [wrapperRef]);

  /**
   * 从左向右滚动
   */
  const gotoNextPage = React.useCallback(() => {
    var _wrapperRef$current;
    const containerRight = ((_wrapperRef$current = wrapperRef.current) === null || _wrapperRef$current === void 0 ? void 0 : _wrapperRef$current.getBoundingClientRect().right) || 0;
    /**
     * 找到第一个未显示完全的 tab
     */
    for (const pane of panes) {
      var _tabElementMapRef$cur;
      const tab = (_tabElementMapRef$cur = tabElementMapRef.current) === null || _tabElementMapRef$cur === void 0 ? void 0 : _tabElementMapRef$cur.get(pane.key);
      if (!tab) {
        continue;
      }
      const {
        right: tabRight
      } = tab.getBoundingClientRect();
      if (tabRight > containerRight) {
        tab.scrollIntoView({
          ...defaultScrollOption,
          // 让选项滚动到最左侧，而不是仅仅滚动到可见区域
          inline: 'start'
        });
        return;
      }
    }
  }, [panes, tabElementMapRef, wrapperRef]);

  /**
   * 从右向左滚动
   */
  const gotoPrePage = React.useCallback(() => {
    var _wrapperRef$current2;
    const containerLeft = ((_wrapperRef$current2 = wrapperRef.current) === null || _wrapperRef$current2 === void 0 ? void 0 : _wrapperRef$current2.getBoundingClientRect().left) || 0;
    /**
     * 找到第一个未显示完全的 tab
     */
    for (const pane of [...panes].reverse()) {
      var _tabElementMapRef$cur2;
      const tab = (_tabElementMapRef$cur2 = tabElementMapRef.current) === null || _tabElementMapRef$cur2 === void 0 ? void 0 : _tabElementMapRef$cur2.get(pane.key);
      if (!tab) {
        continue;
      }
      const {
        left: tabLeft
      } = tab.getBoundingClientRect();
      if (tabLeft < containerLeft) {
        tab.scrollIntoView({
          ...defaultScrollOption,
          // 让选项滚动到最右侧，而不是仅仅滚动到可见区域
          inline: 'end'
        });
        return;
      }
    }
  }, [panes, tabElementMapRef, wrapperRef]);
  const getScrollButtonSx = position => [type !== 'capsule' && {
    "props": {
      "mc9l5x": ukchsq7,
      "qb2dma": ug4hjch,
      "Bt984gj": ubyhozq,
      "a9b677": uevmt6d,
      "Be2twd7": u283mfx,
      "Bceei9c": u12o6wmt,
      "sj55zd": u1u0vkcg,
      "qhf8xq": u5jxv71,
      "ayd6f0": u16ye362
    },
    "alias": rootCssScope('__scroll-button')
  }, (position === 'left' ? scrollButtonDisabled[0] : scrollButtonDisabled[1]) && {
    "props": {
      "Bceei9c": u1vjduc2,
      "sj55zd": u14cxjev
    }
  }];

  /**
   * 选中 tab,将 tabBar 滚动到可视区域内
   */
  const scrollToSelectedTab = React.useCallback(key => {
    var _tabElementMapRef$cur3;
    const tab = (_tabElementMapRef$cur3 = tabElementMapRef.current) === null || _tabElementMapRef$cur3 === void 0 ? void 0 : _tabElementMapRef$cur3.get(key);
    const wrapperEle = wrapperRef.current;
    if (!tab || !wrapperEle) {
      return;
    }
    const {
      right: containerRight,
      left: containerLeft
    } = wrapperEle.getBoundingClientRect();
    const {
      right: tabRight,
      left: tabLeft
    } = tab.getBoundingClientRect();

    // 在左侧
    if (tabLeft < containerLeft + scrollMaskWidth) {
      wrapperEle.scroll({
        // 此时可能出现负数，但浏览器能兼容
        left: tab.offsetLeft - scrollMaskWidth,
        ...defaultScrollOption
      });
      return;
    }

    // 在右侧
    if (tabRight > containerRight - scrollMaskWidth) {
      wrapperEle.scroll({
        left: wrapperEle.scrollLeft + (tabRight - containerRight) + scrollMaskWidth,
        ...defaultScrollOption
      });
    }
  }, [scrollMaskWidth, tabElementMapRef, wrapperRef]);
  return jsx("div", {
    style: {
      columnGap: scrollButtonGutter
    },
    sx: [{
      "props": {
        "mc9l5x": ukchsq7,
        "Bt984gj": ubyhozq,
        "a9b677": u13d3rh8
      }
    }]
  }, Boolean(scrollButtonVisible) && jsx("div", {
    sx: getScrollButtonSx('left'),
    onClick: scrollButtonDisabled[0] ? undefined : gotoPrePage
  }, jsx(IconButton, {
    size: "extra-small",
    disabled: scrollButtonDisabled[0],
    icon: jsx(LeftOutlined, null)
  })), jsx("div", {
    style: {
      '--scroll-mask-width': scrollMaskWidth + 'px'
    },
    sx: [{
      "props": {
        "qhf8xq": u5jxv71,
        "Bf4jedk": urmav4p,
        "Bh6795r": u99gbyh,
        "Bnnss6s": uldrd5o
      }
    }, !scrollButtonDisabled[0] && {
      "props": {
        "rurcny": u1klsf9c,
        "xx9plb": uua3xia,
        "iyvzjs": u16zaoyg,
        "Bab4cu6": u19z4mkn,
        "Byque4d": u1pid3wf,
        "Bf8kmfk": ucgpi4a,
        "Bj2wrql": upyoa5l,
        "Bdn98qo": u102c4m8,
        "pmf9yy": u1mp4r5t,
        "B4zgs9e": uc4om0q
      }
    }, !scrollButtonDisabled[1] && {
      "props": {
        "Bs6t6z0": u1xegxa0,
        "Hdbjpj": u5qwpn8,
        "Bmgsduf": u1l27p0d,
        "Bunff0j": ujik7se,
        "cnmfks": u1yoqygm,
        "u6d25": u1anwibm,
        "Bj55yzk": u1zgj7v,
        "Biobvvw": u1w1hd4l,
        "Fdvyjd": u1aoskr8,
        "Bpudnjh": ubm18w5
      }
    }]
  }, jsx("div", {
    ref: composedRef,
    className: rootCssScope('__tab-bar-wrap'),
    onScroll: checkScrollButtonStatus,
    sx: [{
      "props": {
        "B68tc82": u1qr8jow,
        "qhf8xq": u5jxv71,
        "B15if2v": u9vy1v2,
        "qgh3wz": u1no85ao
      }
    }]
  }, jsx(TabsInkBar, {
    offsetY: inkOffsetY
  }), jsx("div", {
    ref: innerListRef,
    sx: [{
      "props": {
        "mc9l5x": ukchsq7,
        "Eh141a": u125g0x8,
        "qhf8xq": u5jxv71,
        "Bnnss6s": uldrd5o,
        "Bh6795r": ug24l6d
      }
    }]
  }, panes.map(pane => jsx("div", {
    key: pane.key,
    style: resolveSizingBehavior(rootTabsSizingBehavior)
  }, jsx(TabProvider, {
    tabKey: pane.key
  }, jsx(Tab, {
    onClick: () => scrollToSelectedTab(pane.key)
  }))))))), Boolean(scrollButtonVisible) && jsx("div", {
    sx: getScrollButtonSx('right'),
    onClick: scrollButtonDisabled[1] ? undefined : gotoNextPage
  }, jsx(IconButton, {
    size: "extra-small",
    disabled: scrollButtonDisabled[1],
    icon: jsx(RightOutlined, null)
  })));
});
