// @universe-design/react/es/tabs/sx.js
//.uvo9xlb{border-top-left-radius:@border-radius-md;}
import u1773ck5 from "./../.stylex/u1773ck5";
//.ugd3wmz{border-top-right-radius:@border-radius-md;}
import upqf1qz from "./../.stylex/upqf1qz";
//.ueep33z{border-bottom-right-radius:@border-radius-md;}
import u1cxaokx from "./../.stylex/u1cxaokx";
//.u3qetca{border-bottom-left-radius:@border-radius-md;}
import uzx399i from "./../.stylex/uzx399i";
//.u1l2lv2i:hover{border-top-left-radius:@border-radius-md;}
import u8lm1xd from "./../.stylex/u8lm1xd";
//.u7l4e48:hover{border-top-right-radius:@border-radius-md;}
import u1fzbsm8 from "./../.stylex/u1fzbsm8";
//.u1p09kk6:hover{border-bottom-right-radius:@border-radius-md;}
import u1x6icpj from "./../.stylex/u1x6icpj";
//.uhijqq1:hover{border-bottom-left-radius:@border-radius-md;}
import uoa5q5 from "./../.stylex/uoa5q5";
//.u1ipgtue{background-color:@udtoken-component-outlined-bg;}
import u1xrgxpq from "./../.stylex/u1xrgxpq";
//.ub99si7{border-top-color:@line-border-component;}
import uwlx52q from "./../.stylex/uwlx52q";
//.u12908r0{border-right-color:@line-border-component;}
import u10ismar from "./../.stylex/u10ismar";
//.u1lgt5kb{border-bottom-color:@line-border-component;}
import u1g1nbrb from "./../.stylex/u1g1nbrb";
//.u3ps6bl{border-left-color:@line-border-component;}
import uia0x1d from "./../.stylex/uia0x1d";
//.u1es3o6n{color:@primary-content-default;}
import uue9938 from "./../.stylex/uue9938";
//.urdqxp0:hover{background-color:var(--udtoken-tabs-view-bg-hover);}
import uf0hpal from "./../.stylex/uf0hpal";
//.u1rvp7gu{color:@text-caption;}
import ueh72r3 from "./../.stylex/ueh72r3";
//.u3rkx2v{border-bottom-right-radius:@border-radius-sm;}
import u19b90lh from "./../.stylex/u19b90lh";
//.u17hi4zz{border-bottom-left-radius:@border-radius-sm;}
import u1vk5ola from "./../.stylex/u1vk5ola";
//.uyxjys6{border-top-color:@line-border-card;}
import u1n7c7ic from "./../.stylex/u1n7c7ic";
//.u1f4hk3w{border-right-color:@line-border-card;}
import ujkvyav from "./../.stylex/ujkvyav";
//.u14wfv0u{border-left-color:@line-border-card;}
import uqbl7tl from "./../.stylex/uqbl7tl";
//.u96dupq{border-top-left-radius:@border-radius-sm;}
import u19ubm5t from "./../.stylex/u19ubm5t";
//.u1msb61c{border-top-right-radius:@border-radius-sm;}
import umy11dx from "./../.stylex/umy11dx";
//.u1b829s6{background-color:@bg-base;}
import ucbcxg8 from "./../.stylex/ucbcxg8";
//.u1p3nwhy{border-top-color:transparent;}
import u1cd5c9u from "./../.stylex/u1cd5c9u";
//.u11589ue{border-right-color:transparent;}
import uda9nr0 from "./../.stylex/uda9nr0";
//.u1q5o8ev{border-bottom-color:transparent;}
import uy9crqm from "./../.stylex/uy9crqm";
//.u1pdflbu{border-left-color:transparent;}
import uptwv6n from "./../.stylex/uptwv6n";
//.u5ogflp{border-top-width:1px;}
import u1cpcpb5 from "./../.stylex/u1cpcpb5";
//.u1hqa2wf{border-right-width:1px;}
import u10qn3go from "./../.stylex/u10qn3go";
//.u1f09k3d{border-bottom-width:1px;}
import uflkg16 from "./../.stylex/uflkg16";
//.uinvdd3{border-left-width:1px;}
import uvn5txf from "./../.stylex/uvn5txf";
//.uzkkow9{border-top-style:solid;}
import us9j13g from "./../.stylex/us9j13g";
//.ucdblym{border-right-style:solid;}
import umt7kug from "./../.stylex/umt7kug";
//.ug706s2{border-bottom-style:solid;}
import uh3lwu4 from "./../.stylex/uh3lwu4";
//.ujik90z{border-left-style:solid;}
import ur61548 from "./../.stylex/ur61548";
//.u1eyxnol:hover:not(.u1eyxnol--disabled){color:@primary-content-default;}
import u1ymqkzg from "./../.stylex/u1ymqkzg";
//.u1kcqot9{padding-top:12px;}
import u1w9jt9n from "./../.stylex/u1w9jt9n";
//.upe6lb7{padding-bottom:12px;}
import u1hdaclu from "./../.stylex/u1hdaclu";
//.u4d9j23{justify-content:center;}
import ut7q8es from "./../.stylex/ut7q8es";
//.uxldao9{height:28px;}
import un6nsv3 from "./../.stylex/un6nsv3";
//.u1sbtcvk{padding-top:5px;}
import u15pdtm from "./../.stylex/u15pdtm";
//.udghr9{padding-bottom:5px;}
import u1y2b6fc from "./../.stylex/u1y2b6fc";
//.u1xp5gbu{padding-top:6px;}
import u1a72m5n from "./../.stylex/u1a72m5n";
//.u1d7kygh{padding-bottom:6px;}
import uvnwla from "./../.stylex/uvnwla";
//.u1d2rq10{height:32px;}
import ujixoer from "./../.stylex/ujixoer";
//.u8ljn23{height:36px;}
import u1i2nmr7 from "./../.stylex/u1i2nmr7";
//.ubhnoac{height:40px;}
import ubr2xhg from "./../.stylex/ubr2xhg";
//.u1g0x7ka{padding-top:0;}
import u1fdqv2r from "./../.stylex/u1fdqv2r";
//.uat0sn4{padding-right:20px;}
import u1oqvdpm from "./../.stylex/u1oqvdpm";
//.u1qch9an{padding-bottom:0;}
import u6qvdto from "./../.stylex/u6qvdto";
//.uekwl8i{padding-left:20px;}
import ub7yml4 from "./../.stylex/ub7yml4";
//.u1lmfglv{margin-top:0px;}
import u1kxgv9m from "./../.stylex/u1kxgv9m";
//.uibjyge{margin-right:0px;}
import unbze3l from "./../.stylex/unbze3l";
//.u1abmfm4{margin-bottom:0px;}
import u10d06an from "./../.stylex/u10d06an";
//.u9yszdx{margin-left:0px;}
import u1w6ojc6 from "./../.stylex/u1w6ojc6";
//.u10pi13n{position:relative;}
import u5jxv71 from "./../.stylex/u5jxv71";
//.u19g0ac{z-index:1;}
import uhppl2f from "./../.stylex/uhppl2f";
//.up9bwmr{padding-top:8px;}
import u1e84ebu from "./../.stylex/u1e84ebu";
//.u150uoa4{padding-bottom:8px;}
import u8ckwbm from "./../.stylex/u8ckwbm";
//.uq1loh5{padding-top:2px;}
import upgelb9 from "./../.stylex/upgelb9";
//.uutqtb8{padding-bottom:2px;}
import u1vsjzop from "./../.stylex/u1vsjzop";
//.u1phki43{margin-right:8px;}
import ujgl7ix from "./../.stylex/ujgl7ix";
//.u1khb0e9{padding-top:3px;}
import ucil735 from "./../.stylex/ucil735";
//.u19lj068{padding-right:8px;}
import u441rjq from "./../.stylex/u441rjq";
//.u1jnq6q7{padding-bottom:3px;}
import uaonily from "./../.stylex/uaonily";
//.u177v4lu{padding-left:8px;}
import um3isuf from "./../.stylex/um3isuf";
//.u10ra9hq{padding-top:4px;}
import un7s8z9 from "./../.stylex/un7s8z9";
//.u11qrl6u{padding-right:12px;}
import u143d86t from "./../.stylex/u143d86t";
//.u1y2xyjm{padding-bottom:4px;}
import u1ibfn25 from "./../.stylex/u1ibfn25";
//.ujlbh76{padding-left:12px;}
import urjm3ap from "./../.stylex/urjm3ap";
//.u1x4n1mu{margin-right:12px;}
import u1z1dmk from "./../.stylex/u1z1dmk";
/**
 * Stylex CSS
 */

import stylex from '@universe-design/stylex';
import { cssTypographyBody0, cssTypographyBody2, cssTypographyTitle4, cssTypographyTitle5, cssTypographyHeadline, cssTypographyCaption0 } from '../stylex';
import { makeStylexStyles } from '@universe-design/stylex/react';

// Tab 的版面样式
export const sxTabTypography = (size, isSelected) => {
  if (size === 'large') {
    if (isSelected) {
      return cssTypographyTitle4;
    }
    return cssTypographyTitle5;
  }
  if (size === 'medium') {
    if (isSelected) {
      return cssTypographyHeadline;
    }
    return cssTypographyBody0;
  }
  if (isSelected) {
    return cssTypographyCaption0;
  }
  return cssTypographyBody2;
};
export const sxTabSize = (type, size) => {
  var _tabSizes$type, _tabSizes$type2;
  const tabSizes = {
    tag: {
      large: {
        "props": {
          "z8tnut": un7s8z9,
          "z189sj": u143d86t,
          "Byoj8tv": u1ibfn25,
          "uwmqm3": urjm3ap,
          "t21cq0": u1z1dmk
        }
      },
      medium: {
        "props": {
          "z8tnut": ucil735,
          "z189sj": u441rjq,
          "Byoj8tv": uaonily,
          "uwmqm3": um3isuf,
          "t21cq0": u1z1dmk
        }
      },
      small: {
        "props": {
          "z8tnut": upgelb9,
          "z189sj": u441rjq,
          "Byoj8tv": u1vsjzop,
          "uwmqm3": um3isuf,
          "t21cq0": ujgl7ix
        }
      }
    },
    line: {
      small: {
        "props": {
          "z8tnut": u1e84ebu,
          "Byoj8tv": u8ckwbm
        }
      }
    },
    card: {
      base: {
        "props": {
          "B6of3ja": u1kxgv9m,
          "t21cq0": unbze3l,
          "jrapky": u10d06an,
          "Frg6f3": u1w6ojc6,
          "qhf8xq": u5jxv71,
          "Bj3rh1h": uhppl2f
        }
      },
      large: {
        "props": {
          "Bqenvij": ubr2xhg,
          "z8tnut": u1fdqv2r,
          "z189sj": u1oqvdpm,
          "Byoj8tv": u6qvdto,
          "uwmqm3": ub7yml4
        }
      },
      medium: {
        "props": {
          "Bqenvij": u1i2nmr7,
          "z8tnut": u1fdqv2r,
          "z189sj": u1oqvdpm,
          "Byoj8tv": u6qvdto,
          "uwmqm3": ub7yml4
        }
      },
      small: {
        "props": {
          "Bqenvij": ujixoer,
          "z8tnut": u1fdqv2r,
          "z189sj": u1oqvdpm,
          "Byoj8tv": u6qvdto,
          "uwmqm3": ub7yml4
        }
      }
    },
    capsule: {
      large: {
        "props": {
          "Bqenvij": u1i2nmr7,
          "z8tnut": u1a72m5n,
          "z189sj": u143d86t,
          "Byoj8tv": uvnwla,
          "uwmqm3": urjm3ap
        }
      },
      medium: {
        "props": {
          "Bqenvij": ujixoer,
          "z8tnut": u15pdtm,
          "z189sj": u143d86t,
          "Byoj8tv": u1y2b6fc,
          "uwmqm3": urjm3ap
        }
      },
      small: {
        "props": {
          "Bqenvij": un6nsv3,
          "z8tnut": un7s8z9,
          "z189sj": u143d86t,
          "Byoj8tv": u1ibfn25,
          "uwmqm3": urjm3ap
        }
      }
    }
  };
  return stylex.compose({
    "props": {
      "z8tnut": u1w9jt9n,
      "Byoj8tv": u1hdaclu,
      "Brf1p80": ut7q8es
    }
  }, (_tabSizes$type = tabSizes[type]) === null || _tabSizes$type === void 0 ? void 0 : _tabSizes$type.base, (_tabSizes$type2 = tabSizes[type]) === null || _tabSizes$type2 === void 0 ? void 0 : _tabSizes$type2[size]);
};
export const useTabStyles = makeStylexStyles(_ref => {
  let {
    prefix
  } = _ref;
  return {
    tabHover: {
      "props": {
        "lf9awn": u1ymqkzg
      }
    },
    cardBase: {
      "props": {
        "B4j52fo": u1cpcpb5,
        "Bekrc4i": u10qn3go,
        "Bn0qgzm": uflkg16,
        "ibv6hh": uvn5txf,
        "icvyot": us9j13g,
        "vrafjx": umt7kug,
        "oivjwe": uh3lwu4,
        "wvpqe5": ur61548
      }
    },
    cardDefault: {
      "props": {
        "g2u3we": u1cd5c9u,
        "h3c5rm": uda9nr0,
        "B9xav0g": uy9crqm,
        "zhjwy3": uptwv6n
      },
      "alias": `${prefix}__tab--card-with-border`
    },
    cardActive: {
      "props": {
        "g2u3we": u1n7c7ic,
        "h3c5rm": ujkvyav,
        "B9xav0g": uy9crqm,
        "zhjwy3": uqbl7tl,
        "Btl43ni": u19ubm5t,
        "B7oj6ja": umy11dx,
        "De3pzq": ucbcxg8,
        "Bj3rh1h": uhppl2f
      },
      "alias": `${prefix}__tab--card-with-border--active`
    },
    capsuleBase: {
      "props": {
        "sj55zd": ueh72r3,
        "B4j52fo": u1cpcpb5,
        "Bekrc4i": u10qn3go,
        "Bn0qgzm": uflkg16,
        "ibv6hh": uvn5txf,
        "icvyot": us9j13g,
        "vrafjx": umt7kug,
        "oivjwe": uh3lwu4,
        "wvpqe5": ur61548,
        "Btl43ni": u19ubm5t,
        "B7oj6ja": umy11dx,
        "Bbmb7ep": u19b90lh,
        "Beyfa6y": u1vk5ola
      }
    },
    capsuleDefault: {
      "props": {
        "g2u3we": u1cd5c9u,
        "h3c5rm": uda9nr0,
        "B9xav0g": uy9crqm,
        "zhjwy3": uptwv6n,
        "Jwef8y": uf0hpal
      },
      "alias": `${prefix}__tab--capsule-with-border`
    },
    capsuleActive: {
      "props": {
        "De3pzq": u1xrgxpq,
        "g2u3we": uwlx52q,
        "h3c5rm": u10ismar,
        "B9xav0g": u1g1nbrb,
        "zhjwy3": uia0x1d,
        "sj55zd": uue9938
      },
      "alias": `${prefix}__tab--capsule-with-border--active`
    }
  };
});
export const useTabsMoreStyles = makeStylexStyles(_ref2 => {
  let {
    prefix
  } = _ref2;
  return {
    capsuleBase: {
      "props": {
        "sj55zd": ueh72r3
      },
      "alias": `${prefix}__tab-bar-more__capsule`
    },
    capsuleHovered: {
      "props": {
        "Jwef8y": uf0hpal,
        "B2zomw9": u8lm1xd,
        "Fgk69s": u1fzbsm8,
        "gu71dq": u1x6icpj,
        "Btmb5he": uoa5q5
      }
    },
    capsulePressed: {
      "props": {
        "De3pzq": u1xrgxpq,
        "B4j52fo": u1cpcpb5,
        "Bekrc4i": u10qn3go,
        "Bn0qgzm": uflkg16,
        "ibv6hh": uvn5txf,
        "icvyot": us9j13g,
        "vrafjx": umt7kug,
        "oivjwe": uh3lwu4,
        "wvpqe5": ur61548,
        "g2u3we": uwlx52q,
        "h3c5rm": u10ismar,
        "B9xav0g": u1g1nbrb,
        "zhjwy3": uia0x1d,
        "Btl43ni": u1773ck5,
        "B7oj6ja": upqf1qz,
        "Bbmb7ep": u1cxaokx,
        "Beyfa6y": uzx399i,
        "sj55zd": uue9938
      },
      "alias": `${prefix}__tab-bar-more-capsule--active`
    }
  };
});
