// @universe-design/react/es/tabs/Tab.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.u11u7vat{min-width:20px;}
import u1n2z6e7 from "./../.stylex/u1n2z6e7";
/* eslint-disable react/no-unknown-property */
/** @jsxRuntime classic */
/** @jsx jsx */
/** @jsxFrag React.Fragment */
import { jsx, useMergeComponentProps } from '@universe-design/stylex/react';
import React from 'react';
import stylex from '@universe-design/stylex';
import clsx from 'clsx';
import { useOverflowContext } from '../overflow/hooks';
import { useTabsContext } from './TabsContext';
import { forwardHTMLRef } from '../util/forwardRef';
import { Text } from '../text';
import { sxTabSize, sxTabTypography, useTabStyles } from './sx';
import { useTabContext } from './TabContext';
export const Tab = props => {
  const {
    size,
    panes,
    rootCssScope,
    type,
    tabBarGutter,
    handleTabClick,
    handleTabDoubleClick,
    activeTabIndex,
    activeKey,
    tabElementMapRef
  } = useTabsContext();
  const tabStyles = useTabStyles({
    prefix: rootCssScope()
  });
  const {
    key
  } = useTabContext();
  const {
    omittedKeys
  } = useOverflowContext();
  const paneProps = panes.find(p => p.key === key);
  const paneIndex = panes.findIndex(p => p.key === key);
  if (!paneProps) {
    return null;
  }
  const {
    tab,
    disabled
  } = paneProps;
  const isActive = paneIndex === activeTabIndex;
  const cls = clsx(rootCssScope('__tab'), rootCssScope(`__tab--${type}`), {
    [rootCssScope('__tab--active')]: isActive,
    [rootCssScope('__tab--divider')]: type === 'card' && (paneIndex < activeTabIndex - 1 || paneIndex > activeTabIndex || paneIndex === activeTabIndex - 1 && omittedKeys.includes(activeKey)),
    [rootCssScope('__tab--disabled')]: disabled
  }, type === 'tag' && {
    [rootCssScope('-sm-border-radius')]: true,
    [rootCssScope('-neutral-background')]: !isActive,
    [rootCssScope('-primary-background')]: isActive,
    [rootCssScope('-color-pri-text')]: isActive
  });
  return jsx("div", {
    ref: node => {
      var _tabElementMapRef$cur;
      (_tabElementMapRef$cur = tabElementMapRef.current) === null || _tabElementMapRef$cur === void 0 ? void 0 : _tabElementMapRef$cur.set(key, node);
    },
    style: {
      ...(type === 'line' ? {
        marginRight: tabBarGutter
      } : {})
    },
    className: cls,
    key: `tab-nav-key-${paneIndex}`,
    onClick: e => {
      var _props$onClick;
      (_props$onClick = props.onClick) === null || _props$onClick === void 0 ? void 0 : _props$onClick.call(props, e);
      handleTabClick(key, e);
    },
    onDoubleClick: e => handleTabDoubleClick(key, e),
    sx: [type !== 'capsule' && tabStyles.tabHover, sxTabTypography(size, isActive), sxTabSize(type, size), type === 'card' && [tabStyles.cardBase, isActive ? tabStyles.cardActive : tabStyles.cardDefault], type === 'capsule' && [tabStyles.capsuleBase, isActive ? tabStyles.capsuleActive : tabStyles.capsuleDefault], props.sx]
  }, tab);
};

// еӣ дёәеӨ–еұӮ TabItem жҳҜ flexпјҢжІЎжі•дҪҝз”Ё TextпјҢз”ЁиҝҷдёӘеҒҡеҚ•иЎҢзңҒз•Ҙ
export const TabContent = forwardHTMLRef((props, ref) => {
  const {
    isOmitted
  } = useTabContext();
  return jsx(Text, _extends({
    as: "span",
    lineClamp: isOmitted ? undefined : 1,
    ref: ref
  }, useMergeComponentProps(props, {
    sx: [{
      "props": {
        "Bf4jedk": u1n2z6e7
      }
    }]
  })));
});
