// @universe-design/react/es/tabs/Tabs.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.u1acs6jw{width:fit-content;}
import u98ahtk from "./../.stylex/u98ahtk";
//.u9w1357{background:@N900-5;}
import u1suju4n from "./../.stylex/u1suju4n";
//.u96dupq{border-top-left-radius:@border-radius-sm;}
import u19ubm5t from "./../.stylex/u19ubm5t";
//.u1msb61c{border-top-right-radius:@border-radius-sm;}
import umy11dx from "./../.stylex/umy11dx";
//.u3rkx2v{border-bottom-right-radius:@border-radius-sm;}
import u19b90lh from "./../.stylex/u19b90lh";
//.u17hi4zz{border-bottom-left-radius:@border-radius-sm;}
import u1vk5ola from "./../.stylex/u1vk5ola";
//.u1xg1ack{position:sticky;}
import umu3nk9 from "./../.stylex/umu3nk9";
//.u113wtx2{top:0px;}
import u1981n0b from "./../.stylex/u1981n0b";
//.u11fbdcc{background-color:@bg-body;}
import u1to0gca from "./../.stylex/u1to0gca";
//.u10pi13n{position:relative;}
import u5jxv71 from "./../.stylex/u5jxv71";
/* eslint-disable react/no-unknown-property */
/** @jsxRuntime classic */
/** @jsx jsx */
/** @jsxFrag React.Fragment */
import stylex from '@universe-design/stylex';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx } from '@universe-design/stylex/react';
import DownBoldOutlined from '@universe-design/icons-react/DownBoldOutlined';
import { cssTypographyTitle4, cssTypographyTitle5 } from '../stylex';
import clsx from 'clsx';
import React, { forwardRef, useImperativeHandle, useMemo, useRef, useState } from 'react';
import { useLocaleReceiver } from '../config-provider/hooks';
import { Dropdown } from '../dropdown';
import { Menu } from '../menu';
import { Overflow } from '../overflow';
import { useOverflowContext } from '../overflow/hooks';
import { getDefaultSizeAbbr, isFunction, __SCREENSHOT__ } from '../util';
import { TabPane } from './TabPane';
import { TabsContext, useTabs, useTabsContext } from './TabsContext';
import { sxTabSize, sxTabTypography, useTabsMoreStyles } from './sx';
import { Tab } from './Tab';
import { TabProvider } from './TabContext';
import { ScrollableTabsHeader } from './ScrollableTabsHeader';
import { TabsInkBar } from './TabInk';
import './style';
import { resolveSizingBehavior } from './sizing_behavior';
export const TabMoreArrow = () => jsx(DownBoldOutlined, null);
export const TabMoreContent = () => {
  const locale = useLocaleReceiver('tabs');
  return jsx("span", null, locale.more);
};

/**
 * `<TabMore />` 组件的属性
 * @public
 */

export const TabMore = _ref => {
  var _ref3;
  let {
    children,
    ...dropDownProps
  } = _ref;
  const {
    panes,
    rootCssScope,
    handleTabClick,
    activeKey,
    moreTabRef,
    size,
    type
  } = useTabsContext();
  const {
    omittedKeys
  } = useOverflowContext();
  const [showMoreTabs, setShowMoreTabs] = useState(false);
  const isSelected = omittedKeys.includes(activeKey);
  const mergedVisible = __SCREENSHOT__ ? isSelected : showMoreTabs && Boolean(omittedKeys.length);
  const tabsMoreStyles = useTabsMoreStyles({
    prefix: rootCssScope()
  });
  const morePanes = panes.filter(p => omittedKeys.includes(p.key));
  const cls = clsx(rootCssScope('__tab-bar-more'), rootCssScope(`__tab-bar-more--${type}`), {
    [rootCssScope('__tab-bar-more--selected')]: isSelected,
    [rootCssScope('__tab-bar-more--show')]: mergedVisible
  });
  const menu = jsx(Menu, {
    selectedKeys: [activeKey],
    textOverflow: false,
    selectable: true
  }, morePanes.map(pane => {
    const {
      disabled,
      key,
      tab
    } = pane;
    return jsx(Menu.Item, {
      key: key,
      disabled: disabled,
      onClick: _ref2 => {
        let {
          domEvent
        } = _ref2;
        return handleTabClick(key, domEvent);
      }
    }, jsx(TabProvider, {
      tabKey: key
    }, tab));
  }));
  const moreTabContentNode = jsx(React.Fragment, null, jsx(TabMoreContent, null), jsx(TabMoreArrow, null));
  return jsx("div", {
    className: clsx(rootCssScope('__tab-bar__operation'))
  }, jsx("div", {
    className: cls,
    sx: [!isSelected ? cssTypographyTitle5 : cssTypographyTitle4]
  }, jsx(Dropdown, _extends({
    trigger: ['hover', 'click']
  }, dropDownProps, {
    overlay: menu,
    visible: mergedVisible,
    onVisibleChange: setShowMoreTabs
  }), jsx("div", {
    ref: moreTabRef,
    className: clsx(rootCssScope('__tab-bar-more-body'), type === 'tag' && [rootCssScope('-sm-border-radius'), isSelected ? [rootCssScope('-primary-background'), rootCssScope('-color-pri-text')] : rootCssScope('-neutral-background')]),
    sx: [sxTabTypography(size, isSelected), sxTabSize(type, size), type === 'capsule' && [tabsMoreStyles.capsuleBase, !isSelected && tabsMoreStyles.capsuleHovered, isSelected && tabsMoreStyles.capsulePressed]]
  }, (_ref3 = isFunction(children) ? children({
    omittedKeys: omittedKeys
  }) : children) !== null && _ref3 !== void 0 ? _ref3 : moreTabContentNode))));
};
export const Tabs = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    tabMore,
    sticky
  } = props;
  const tabCtx = useTabs(props);
  const {
    style,
    className,
    tabBarStyle,
    type = 'line',
    size = 'medium',
    maxCount,
    tabBarExtraContent,
    rootCssScope,
    tabBarWrapperRef,
    panes,
    rootTabsSizingBehavior,
    scrollable
  } = tabCtx;
  const usedSize = getDefaultSizeAbbr(size, 'medium');
  const tabCls = clsx(rootCssScope(), {
    [rootCssScope(`--type-${type}`)]: type,
    [rootCssScope(`--size-${usedSize}`)]: usedSize
  }, className);
  const tabsRef = useRef(null);
  // 蓝色选中线和 divider 重合
  const inkOffsetY = 0;
  useImperativeHandle(ref, () => tabsRef.current);
  const tabsInkBarNode = useMemo(() => jsx(TabsInkBar, {
    offsetY: inkOffsetY
  }), [inkOffsetY]);
  const moreTabNode = useMemo(() => tabMore !== null && tabMore !== void 0 ? tabMore : jsx(TabMore, null), [tabMore]);
  const renderLeftExtraContentNode = useMemo(() => {
    const assertExtra = tabBarExtraContent;
    if (!assertExtra) {
      return null;
    }
    return jsx("div", {
      className: rootCssScope('__tab-extra-content')
    }, assertExtra.left);
  }, [rootCssScope, tabBarExtraContent]);
  const renderRightExtraContentNode = useMemo(() => {
    const assertExtra = tabBarExtraContent;
    if (!assertExtra) {
      return null;
    }
    return jsx("div", {
      className: rootCssScope('__tab-extra-content')
    }, assertExtra.right || !assertExtra.left && assertExtra || null);
  }, [rootCssScope, tabBarExtraContent]);
  const getItemStyle = React.useCallback(() => resolveSizingBehavior(rootTabsSizingBehavior), [rootTabsSizingBehavior]);
  return jsx(TabsContext.Provider, {
    value: tabCtx
  }, jsx("div", {
    ref: tabsRef,
    style: style,
    className: tabCls
  }, (() => jsx("div", {
    className: clsx(rootCssScope('__tab-bar-holder'), rootCssScope('__tab-bar-nav'), sticky && rootCssScope('__tab-bar-nav--sticky')),
    sx: [{
      "props": {
        "qhf8xq": u5jxv71
      }
    }, sticky && {
      "props": {
        "qhf8xq": umu3nk9,
        "Bhzewxz": u1981n0b,
        "De3pzq": u1to0gca
      }
    }, type === 'capsule' && {
      "props": {
        "ayd6f0": u1suju4n,
        "Btl43ni": u19ubm5t,
        "B7oj6ja": umy11dx,
        "Bbmb7ep": u19b90lh,
        "Beyfa6y": u1vk5ola
      }
    },
    // scrollable 时会撑开父级，无法滚动 && responsive 时无法正常展开，只有一个 more
    type === 'capsule' && !scrollable && maxCount !== 'responsive' && {
      "props": {
        "a9b677": u98ahtk
      }
    }]
  }, renderLeftExtraContentNode, scrollable ? jsx(ScrollableTabsHeader, {
    ref: tabBarWrapperRef,
    inkOffsetY: inkOffsetY
  }) : jsx("div", {
    ref: tabBarWrapperRef,
    className: rootCssScope('__tab-bar-wrap')
  }, jsx(Overflow, {
    style: tabBarStyle,
    getItemStyle: getItemStyle,
    restStyle: resolveSizingBehavior(rootTabsSizingBehavior),
    keys: panes.map(p => p.key),
    maxCount: maxCount,
    itemRenderer: key => jsx(TabProvider, {
      tabKey: key
    }, jsx(Tab, null)),
    restRenderer: () => moreTabNode,
    addonRenderer: () => tabsInkBarNode,
    alwaysObserveResize: true
  })), renderRightExtraContentNode, type === 'line' && jsx("div", {
    className: rootCssScope('__tab-bar-holder-divider')
  })))(), (() => jsx("div", {
    className: rootCssScope('__content')
  }, panes.map(paneProps => jsx(TabPane, _extends({}, paneProps, {
    internalKey: paneProps.key,
    key: paneProps.key
  })))))()));
});
if (process.env.NODE_ENV !== "production") {
  Tabs.displayName = 'Tabs';
}
