// @universe-design/react/es/tabs/TabsContext.js
import * as React from 'react';
import { createContext, useContext } from 'use-context-selector';
import { useControlledState } from '@universe-design/react-utilities';
import { useCssScope } from '../config-provider/hooks';
import { isFunction, toFlatArray } from '../util';

/**
 * Tab 尺寸行为类型
 * 定义单个 Tab 的尺寸控制方式
 * 支持简单字符串类型和详细的 SizingBehavior 对象类型
 */

export const UD_TABPANE_ID = 'UD_TABPANE_ID';
export function isTabPane(node) {
  var _node$type;
  return node === null || node === void 0 ? void 0 : (_node$type = node.type) === null || _node$type === void 0 ? void 0 : _node$type[UD_TABPANE_ID];
}

/**
 * @public
 */

// 'editable-card' 类型暂缓提供

/**
 * `<Tabs />` 组件的属性
 * @public
 */

/**
 * `<Tabs.Pane>` 组件的属性
 * @public
 */

export const TabsContext = createContext({});
if (process.env.NODE_ENV !== "production") {
  TabsContext.displayName = 'TabsContext';
}
export const useTabs = props => {
  const {
    style,
    className,
    children,
    type = 'line',
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    tabBarGutter = type === 'tag' ? 16 : 28,
    scrollButtonGutter = type === 'line' ? 28 : 0,
    tabBarStyle,
    activeKey: actKeyInProps,
    defaultActiveKey,
    size = 'medium',
    maxCount: maxCountProp,
    tabBarExtraContent,
    onChange,
    onTabClick,
    onTabDoubleClick,
    scrollable: scrollableProp,
    tabsSizingBehavior
  } = props;
  const maxCount = (() => {
    if (!maxCountProp && type === 'capsule') {
      return Infinity;
    }
    if (!maxCountProp) {
      return 'responsive';
    }
    return maxCountProp;
  })();
  const scrollable = (() => // if (type === 'capsule') {
  //   warning(!scrollableProp, devWarning('Tabs', 'scrollable is not supported in capsule type'));
  //   return false;
  // }
  scrollableProp !== null && scrollableProp !== void 0 ? scrollableProp : false)();
  const rootTabsSizingBehavior = React.useMemo(() => {
    if (!tabsSizingBehavior || tabsSizingBehavior === 'hug') {
      return {
        mode: 'hug',
        maxWidth: 240
      };
    }
    if (tabsSizingBehavior === 'flex') {
      return {
        mode: 'flex',
        maxWidth: 240
      };
    }
    return tabsSizingBehavior;
  }, [tabsSizingBehavior]);
  const panes = React.useMemo(() => toFlatArray(children).filter(isTabPane).map((child, index) => ({
    ...child.props,
    key: child.key && String(child.key) || `${index}`
  })), [children]);
  const [activeKey, setActiveKey] = useControlledState(defaultActiveKey || panes[0].key, actKeyInProps);
  const handleTabClick = React.useCallback((key, e) => {
    const pane = panes.find(p => p.key === key);
    if (!pane || pane.disabled) {
      return;
    }
    setActiveKey(key);
    isFunction(onChange) && onChange(key);
    isFunction(onTabClick) && onTabClick(key, e);
  }, [onChange, onTabClick, panes, setActiveKey]);
  const handleTabDoubleClick = React.useCallback((key, e) => {
    const pane = panes.find(p => p.key === key);
    if (!pane || pane.disabled) {
      return;
    }
    if (isFunction(onTabDoubleClick)) {
      onTabDoubleClick(key, e);
    }
  }, [onTabDoubleClick, panes]);
  const rootCssScope = useCssScope('tabs');
  const activeTabIndex = panes.findIndex(p => p.key === activeKey);
  const tabElementMapRef = React.useRef(new Map());
  // position relative 的元素，用于定位 ink
  const tabBarWrapperRef = React.useRef(null);
  // 更多 tab
  const moreTabRef = React.useRef(null);
  return React.useMemo(() => ({
    style,
    className,
    children,
    tabBarGutter,
    tabBarStyle: {
      // avoid browser's text selection behavior, when double click
      userSelect: isFunction(onTabDoubleClick) ? 'none' : 'auto',
      ...tabBarStyle,
      ...(scrollable ? {
        flexWrap: 'nowrap'
      } : {})
    },
    activeKey,
    type,
    size,
    maxCount,
    tabBarExtraContent,
    onChange,
    onTabClick,
    onTabDoubleClick,
    rootCssScope,
    setActiveKey,
    handleTabClick,
    handleTabDoubleClick,
    panes,
    activeTabIndex,
    tabElementMapRef,
    tabBarWrapperRef,
    moreTabRef,
    scrollButtonGutter,
    rootTabsSizingBehavior,
    scrollable
  }), [activeKey, activeTabIndex, children, className, handleTabClick, handleTabDoubleClick, maxCount, onChange, onTabClick, onTabDoubleClick, panes, rootCssScope, scrollButtonGutter, scrollable, setActiveKey, size, style, tabBarExtraContent, tabBarGutter, tabBarStyle, type, rootTabsSizingBehavior]);
};
export const useTabsContext = () => useContext(TabsContext);
