// @universe-design/react/src/portal/index.tsx
import * as React from 'react';
import ReactDOM from 'react-dom';
import clsx from 'clsx';
import { isFunction } from '../util';
import { PortalContext, usePortalContext } from './portal-context';
import { useCssScope } from '../config-provider/hooks';
import type { ThunkElement } from '../util/getThunkElement';
import { getThunkElement } from '../util/getThunkElement';

const { Provider: PortalContextProvider } = PortalContext;

export type Container = ThunkElement<Element>;

interface NonPortalProps {
  container: false;
  wrapperClassName?: string; // 没用，只是为了某些地方少判断一点
}

interface DefaultPortalProps {
  container?: null;
  wrapperClassName?: string;
  appendToParentPortal?: boolean;
  parentPortal?: HTMLDivElement | null;
  iframeRender?: boolean;
}

interface ContainerPortalProps {
  iframeRender?: boolean;
  container: NonNullable<Container>;
  wrapperClassName?: string;
}

export type PortalProps = (DefaultPortalProps | ContainerPortalProps | NonPortalProps) & {
  children?: React.ReactNode;
};

abstract class PortalBase<T extends PortalProps> extends React.Component<
  T & {
    innerRef?: React.ForwardedRef<HTMLDivElement>;
    innerWrapperCls: string;
  },
  never
> {
  protected currentHost: Element | null = null;
  protected templateNode: HTMLSpanElement | null = null;
  private currentDocument: Document | null = null;
  private portal: HTMLDivElement | null = null;
  private currentWrapperCls = '';

  protected firstRender = true;

  abstract getHost: (hostOnScreen?: Element | null, templateNode?: HTMLSpanElement | null) => Element | null;

  componentDidMount() {
    this.upsertHost();
    this.updateCls();
  }

  componentDidUpdate() {
    this.upsertHost();
    this.updateCls();
  }

  updateRef = () => {
    const { innerRef } = this.props;
    innerRef &&
      ((isFunction(innerRef) && innerRef(this.portal)) ||
        (typeof innerRef === 'object' && (innerRef.current = this.portal)));
  };

  updateCls = () => {
    const { innerWrapperCls, wrapperClassName } = this.props;
    if (this.currentWrapperCls !== wrapperClassName) {
      this.currentWrapperCls = wrapperClassName || '';
      this.portal && (this.portal.className = clsx(innerWrapperCls, wrapperClassName));
    }
  };

  innerGetHost = () => this.getHost(this.currentHost, this.templateNode);

  upsertHost = () => {
    const nextHost = this.innerGetHost();

    if (nextHost !== this.currentHost) {
      const hostOnScreen = this.currentHost;
      this.currentHost = nextHost;

      if (hostOnScreen && this.portal) {
        // 从当前页面上存在的 host 里卸载之前的 portal
        hostOnScreen.contains(this.portal) && hostOnScreen.removeChild(this.portal);
      }

      if (!this.currentHost) {
        return;
      }

      const nextDocument = this.currentHost.ownerDocument;
      if (nextDocument && this.currentDocument !== nextDocument) {
        this.portal = nextDocument.createElement('div');
        this.currentDocument = nextDocument;
        // 初次渲染的时候可以不用执行 setState
        !this.firstRender && this.setState({});
        this.updateRef();
      }

      if (this.portal) {
        this.currentHost.appendChild(this.portal);
      }
    }

    this.firstRender && (this.firstRender = false);
  };

  componentWillUnmount() {
    if (this.currentHost && this.portal) {
      this.currentHost.contains(this.portal) && this.currentHost.removeChild(this.portal);
    }
    this.portal = null;
    this.currentHost = null;
    this.currentDocument = null;
    this.templateNode = null;
  }

  render() {
    const { children } = this.props;
    this.upsertHost();
    return this.portal ? (
      ReactDOM.createPortal(
        <PortalContextProvider value={this.portal}>{children}</PortalContextProvider>,
        this.portal,
        // span 的 ref 中只记录, 不清除 templateNode,
        // 是为了如果把 children 渲染出来了此时 span 没了
        // 但是会触发 componentDidUpdate
        // componentDidUpdate 里触发 getHost, 此时如果 templateNode 随着 span 消失而变成 null 的话
        // 那前后两次的 host 可能就不一样了
      )
    ) : (
      <span ref={n => n && (this.templateNode = n)} />
    );
  }
}

class ContainerPortal extends PortalBase<ContainerPortalProps> {
  getHost = () => {
    const { container } = this.props;
    return getThunkElement(container);
  };
}

class DefaultPortal extends PortalBase<DefaultPortalProps> {
  getHost = (currentHost?: Element | null, templateNode?: HTMLSpanElement | null) => {
    const { appendToParentPortal = true, parentPortal = null, iframeRender = false } = this.props;
    if (!currentHost && (iframeRender || (appendToParentPortal && parentPortal)) && !templateNode) {
      /**
       * 几种情况需要让 host 是 null
       * 1. 当 iframeRender 为 false 时, 表示不用考虑渲染在 iframe 的场景
       *    因此这种情况下可以在第一次渲染的时候就直接拿到精确的 host
       *    但是在嵌套的场景中, 如果允许把 portal 渲染在父级 portal 上的话
       *    由于在渲染父级 portal 时已经把 context 中注入了 parentPortal
       *    所以这里若是不直接让 host = null 的话
       *    会导致之后直接走到 appendChild 的逻辑
       *    此时万一这个子 portal 有个哥哥节点的话
       *    那这个 portal 就会比它哥哥先被挂到父 portal 下, 顺序就乱了
       *
       * 2. 用户希望可以在 iframe 中渲染, 并且此时还没有临时的 span, 也就意味着此时无论如何也拿不到精准的 document
       */
      return null;
    }

    if (appendToParentPortal && parentPortal) {
      return parentPortal;
    }

    if (iframeRender && templateNode) {
      return templateNode.ownerDocument.body;
    }

    return document.body;
  };
}

export const Portal = React.forwardRef<HTMLDivElement, PortalProps>((props, ref) => {
  const parentPortal = usePortalContext();
  const wrapperClsScope = useCssScope('portal')();

  if (props.container === false) {
    return <>{props.children}</>;
  }

  return props.container !== undefined && props.container !== null ? (
    <ContainerPortal {...props} container={props.container} innerRef={ref} innerWrapperCls={wrapperClsScope} />
  ) : (
    <DefaultPortal {...props} parentPortal={parentPortal} innerWrapperCls={wrapperClsScope} innerRef={ref} />
  );
});
if (__DEV__) {
  Portal.displayName = 'Portal';
}

export default Portal;
