// @universe-design/react/src/context-menu/index.tsx
import type { FC } from 'react';
import React from 'react';
import ReactDOM from 'react-dom';
import { useComplementEvent } from '../hooks/useComplementEvent';
import { useCssScope } from '../config-provider/hooks';
import { isFunction } from '../util';
import { useIsomorphicLayoutEffect as useLayoutEffect } from '../hooks';
import { CrossRootConfigProvider } from '../config-provider';
import { DropdownMenuContext } from '../dropdown/DropdownMenuContext';

import './style';

export interface ContextMenuOption {
  menu: (() => React.ReactElement) | React.ReactElement;
  position: [number, number];
  getPopupContainer?: (triggerNode?: HTMLElement) => HTMLElement;
  onClose?: () => void;
}

interface ContextMenuProps {
  position: [number, number];
  overlay: (() => React.ReactElement) | React.ReactElement;
  onClose?: () => void;
}

const ContextMenu: FC<ContextMenuProps> = props => {
  const { position, onClose, overlay } = props;
  const ref = React.useRef<HTMLDivElement>(null);
  const containerRef = React.useRef<HTMLDivElement>(null);
  const cssScope = useCssScope('context-menu');
  // 点击非menu区域后关闭
  useComplementEvent('click', {
    listener: React.useCallback(() => {
      onClose && onClose();
    }, [onClose]),
    elements: React.useMemo(() => [ref], []),
  });

  useLayoutEffect(() => {
    if (containerRef.current && ref.current) {
      // const Rect;
      const [left, top] = position;
      const result = { left, top };
      const el = ref.current;
      const cel = containerRef.current;
      const height = el.clientHeight;
      const width = el.clientWidth;

      // console.log('el', { left, top, height, width });
      // console.log('cel', { height: cel.clientHeight, width: cel.clientWidth });

      if (height + top > cel.clientHeight) {
        result.top = result.top - height;
        if (result.top < 0) {
          result.top = 0;
        }
      }
      if (width + left > cel.clientWidth) {
        result.left = left - width;
        if (result.left < 0) {
          result.left = 0;
        }
      }
      ref.current.style.left = `${result.left}px`;
      ref.current.style.top = `${result.top}px`;
    }
  }, [position]);

  const overlayElement = React.useMemo(
    (): React.ReactElement => (isFunction(overlay) ? overlay() : overlay),
    [overlay],
  );

  return (
    // 为了告诉 menu 这是一个 dropdown
    <DropdownMenuContext.Provider value={React.useMemo(() => ({}), [])}>
      <div ref={containerRef} className={cssScope('__wrap')} onContextMenu={e => e.stopPropagation()}>
        <div ref={ref} className={cssScope('__overlay')} onClick={onClose}>
          {overlayElement}
        </div>
      </div>
    </DropdownMenuContext.Provider>
  );
};

export const contextMenu = {
  open: (option: ContextMenuOption) => {
    const { menu, position, getPopupContainer, onClose } = option;

    // const uuid = `ud-confirm-instance-${new Date().getTime()}`;

    const portalEl = getPopupContainer ? getPopupContainer() : document.body;
    const container = document.createElement('div');
    portalEl.appendChild(container);

    const destroy = () => {
      const unmountResult = ReactDOM.unmountComponentAtNode(container);
      if (unmountResult && container.parentNode) {
        container.parentNode.removeChild(container);
        onClose && onClose();
      }
    };

    ReactDOM.render(
      <CrossRootConfigProvider>
        <ContextMenu overlay={menu} position={position} onClose={destroy} />
      </CrossRootConfigProvider>,
      container,
    );
  },
};
