// @universe-design/react/src/shared/components/popper/imperative.stories.tsx
import type { Meta } from '@storybook/react';
import { popoverCss } from '../../../_stories/common/constants';
import * as React from 'react';
import { Space, Button } from '../../..';
import type { ImperativePopperController, ImperativePopperControllerProps } from './imperative';
import { popper } from './imperative';

export default {
  title: 'Shared Components/Popper/Imperative',
} satisfies Meta;

export const ImperativePopperBase = () => (
  <Button
    onClick={ev => {
      popper.create({
        defaultVisible: true,
        popup: <div className={popoverCss}>popup</div>,
        reference: ev.currentTarget,
      });
    }}
  >
    Open Popper
  </Button>
);

export const ImperativePopperRecursive = () => {
  const PopperButton = React.useMemo(
    () => () =>
      (
        <Button
          onClick={ev => {
            popper.create({
              defaultVisible: true,
              popup: (
                <div className={popoverCss}>
                  popup <PopperButton />
                </div>
              ),
              reference: ev.currentTarget,
            });
          }}
        >
          Open Popper
        </Button>
      ),
    [],
  );
  return <PopperButton />;
};

export const ImperativePopperStandalone = () => {
  const popperInstRef = React.useRef<ImperativePopperController>();
  const handleClick = React.useCallback<React.MouseEventHandler<HTMLDivElement>>(ev => {
    const sharedProps: ImperativePopperControllerProps = {
      destroyOnHide: false,
      popup: <div className={popoverCss}>popup {ev.currentTarget.textContent}</div>,
      reference: ev.currentTarget,
    };
    if (!popperInstRef.current) {
      popperInstRef.current = popper.create({
        ...sharedProps,
        visible: true,
      });
    }
    const popperInst = popperInstRef.current;
    popperInst.update({
      ...sharedProps,
      visible: true,
      onVisibleChange: visible => {
        popperInst.update({
          ...sharedProps,
          visible,
        });
      },
    });
  }, []);
  return (
    <Space>
      <Button onClick={handleClick}>Button 1</Button>
      <Button onClick={handleClick}>Button 2</Button>
      <Button onClick={handleClick}>Button 3</Button>
    </Space>
  );
};

export const ImperativePopperVirtualElement = () => {
  const containerRef = React.useRef<HTMLDivElement>(null);
  return (
    <div
      ref={containerRef}
      style={{
        display: 'inline-flex',
        position: 'relative',
        overflow: 'hidden',
      }}
    >
      <div
        style={{
          background: 'red',
          height: 300,
          width: 300,
        }}
        onClick={ev => {
          const mouseRect = new DOMRect(ev.clientX, ev.clientY, 0, 0);
          // 提供 contextElement 的时候可以通过 popupPortalContainer 来进行 boundary 判定
          popper.create({
            popupPortalContainer: containerRef,
            defaultVisible: true,
            popup: <div className={popoverCss}>popup</div>,
            reference: {
              contextElement: ev.currentTarget,
              getBoundingClientRect: () => mouseRect,
            },
          });
        }}
      >
        Open Popper
      </div>
      <div
        style={{
          background: 'green',
          height: 300,
          width: 300,
        }}
        onClick={ev => {
          const mouseRect = new DOMRect(ev.clientX, ev.clientY, 0, 0);
          // 不提供 contextElement 的时候可以通过指定 boundary 来进行 boundary 判定
          popper.create({
            boundary: containerRef.current ?? undefined,
            defaultVisible: true,
            popup: <div className={popoverCss}>popup</div>,
            reference: {
              getBoundingClientRect: () => mouseRect,
            },
          });
        }}
      >
        Open Popper with no contextElement
      </div>
    </div>
  );
};
