// @universe-design/react/src/shared/components/popper/imperative.tsx
import type { VirtualElement } from '@floating-ui/core';
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import type { PopperProps } from '.';
import { Popper } from './Popper';
import { CrossRootConfigProvider } from '../../../config-provider';
import { isFunction } from '../../../util';

const ImperativePopperReference = React.forwardRef(
  (props: { reference: Element | VirtualElement }, ref: React.Ref<Element | VirtualElement>) => {
    React.useImperativeHandle(ref, () => props.reference);
    return <></>;
  },
);

interface ImperativePopperRef {
  update: (props: PopperProps) => void;
}

interface ImperativePopperProps {
  defaultProps: PopperProps;
}

const ImperativePopper = React.forwardRef<ImperativePopperRef, ImperativePopperProps>(({ defaultProps }, ref) => {
  const [props, setProps] = React.useState(defaultProps);
  React.useImperativeHandle(
    ref,
    () => ({
      update: nextProps => {
        setProps(nextProps);
      },
    }),
    [],
  );
  return <Popper {...props} />;
});

export interface ImperativePopperControllerProps
  extends Omit<PopperProps, 'children' | 'trigger' | 'showAction' | 'hideAction' | 'alignPoint'> {
  reference: Element | VirtualElement;
}

export class ImperativePopperController {
  private _popperManager: PopperManager;
  // 用做 ReactDOM render 的 container
  private _container: HTMLDivElement | undefined;
  private _imperativePopperRef = React.createRef<ImperativePopperRef>();
  private _isAlive = false;

  constructor(props: ImperativePopperControllerProps, popper: PopperManager) {
    this._popperManager = popper;
    this._initOrUpdate(props);
  }

  update = (props: ImperativePopperControllerProps) => {
    this._initOrUpdate(props);
  };

  destroy = () => {
    this._cleaup();
  };

  private props2PopperProps(props: ImperativePopperControllerProps): PopperProps {
    const { reference, destroyOnHide = true, onAfterClose, ...popperProps } = props;
    return {
      // defaultVisible: true,
      popupPortalContainer: false, // 默认在 container 原位渲染
      ...popperProps,
      destroyOnHide,
      onAfterClose: () => {
        isFunction(onAfterClose) && onAfterClose();
        if (destroyOnHide) {
          this._cleaup();
        }
      },
      children: <ImperativePopperReference reference={reference} />,
      domOnly: false, // 为了让 virtual element 生效需要使用 domOnly false
    };
  }

  private _initOrUpdate = (props: ImperativePopperControllerProps) => {
    if (!this._isAlive) {
      // console.log('_init', this, props);
      this._isAlive = true;
      this._container = document.createElement('div');
      document.body.appendChild(this._container);
      ReactDOM.render(
        <CrossRootConfigProvider>
          <ImperativePopper ref={this._imperativePopperRef} defaultProps={this.props2PopperProps(props)} />
        </CrossRootConfigProvider>,
        this._container,
      );
      this._popperManager.controllers.add(this);
      return;
    }
    // console.log('_update', this, props);
    this._imperativePopperRef.current?.update(this.props2PopperProps(props));
  };

  private _cleaup = () => {
    this._popperManager.controllers.delete(this);
    if (this._container) {
      ReactDOM.unmountComponentAtNode(this._container);
      document.body.removeChild(this._container);
    }
    this._isAlive = false;
  };
}

class PopperManager {
  controllers: Set<ImperativePopperController> = new Set();

  create = (props: ImperativePopperControllerProps) => new ImperativePopperController(props, this);
}

export const popper = new PopperManager();
