// @universe-design/react/src/shared/components/popper/index.stories.tsx
import type { Meta } from '@storybook/react';
import { PerformanceContent } from '../../../_stories/common/PerformanceContent';
import { popoverCss } from '../../../_stories/common/constants';
import React from 'react';
import type { FormContextType, PositionType } from '../../..';
import { useLogger } from '@universe-design/react-utilities';
import {
  Space,
  Button as _Button,
  Form,
  Checkbox,
  FormSelector,
  Switch,
  Select,
  PositionEnum,
  InputNumber,
  ConfigProvider,
  Dropdown,
  FloatingUIPopperSidecar,
} from '../../..';
import { Popover as OldPopper } from '../../../popover';
import type { Placement } from './types';
import { SlideUpDownTransition } from '../transition/SlideUpDownTransition';
import type { PopperProps } from '.';
import { Popper, PopperArrow, PopperBoundary } from '.';
// import { useProxyRef } from '@universe-design/react-utilities';

import './index.stories.less';
import { getPlacements } from '../..';

export default {
  title: 'Shared Components/Popper',
  component: Popper,
} satisfies Meta;

const buttonWidth = 80;
const spaceSize = 24;

const Button = React.memo(_Button);
// const Button = _Button;

const RecursizePopover = ({ prefix = '' }: { prefix?: string }) => (
  <Space direction="vertical">
    {!prefix.match('r-r-r')
      ? [0, 1, 2].map(o => {
          const trigger = o === 0 ? 'click' : o === 1 ? 'hover' : 'contextMenu';
          return (
            <Popper
              key={o}
              offset={10}
              popup={
                <div className={popoverCss}>
                  <RecursizePopover prefix={`r-${prefix}=${o}`} />
                </div>
              }
              trigger={trigger}
              enableMaxHeight
            >
              <Button id={`${prefix}-${trigger}-${o}`}>{`${prefix}-${trigger}-${o}`}</Button>
            </Popper>
          );
        })
      : null}
  </Space>
);

const Log = ({ name }: { name: string }) => {
  useLogger(name);
  return null;
};

const Tooltip = ({ placement }: { placement: Placement }) => {
  const [vis, setVis] = React.useState(false);
  React.useLayoutEffect(() => {
    console.log('Tooltip.useLayoutEffect', document.getElementById('gux-click-0'));
    Promise.resolve().then(() => {
      console.log('Promise Tooltip.useLayoutEffect', document.getElementById('gux-click-0'));
    });
  });
  return (
    <Popper
      // visible={placement === 'bottomLeft' || placement === 'bottomRight'}
      visible={vis}
      onVisibleChange={setVis}
      arrow={<PopperArrow hasStroke />}
      arrowPadding={8 + 8}
      offset={4 + 6}
      placement={placement}
      popup={
        <div className={popoverCss}>
          <div style={{ overflowWrap: 'anywhere' }}>
            <Log name={`${placement} tooltip child`} />
            <RecursizePopover prefix="gux" />
            <PerformanceContent
              componentName={`Popup ${placement}`}
              count={1000}
              itemRenderer={i =>
                placement === 'bottom-start' ? (
                  <Popper arrow={<PopperArrow />} popup={placement.repeat(1000)} enableMaxHeight>
                    <div>{placement}</div>
                  </Popper>
                ) : (
                  <OldPopper
                    // showAction={['click', 'contextMenu', 'mouseEnter']}
                    // hideAction={['click', 'mouseLeave']}
                    content={placement.repeat(1000)}
                  >
                    <div>{placement}</div>
                  </OldPopper>
                )
              }
            />
          </div>
        </div>
      }
      trigger={['click']}
      // enableMaxHeight
      // matchReferenceWidth
      Transition={SlideUpDownTransition}
    >
      <Button style={{ width: buttonWidth }}>{placement}</Button>
    </Popper>
  );
};

export const Basic = () => (
  <PopperBoundary
    id="container"
    style={{
      padding: '48px 0 0 200px',
      height: 600,
      overflow: 'auto',
      border: '2px dashed #ff6b81',
      background: '#281e36',
    }}
  >
    <div style={{ height: 1200 }} />
    <Space
      size={spaceSize}
      style={{
        display: 'flex',
        marginLeft: buttonWidth + spaceSize,
      }}
    >
      <Tooltip placement="top-start" />
      <Tooltip placement="top" />
      <Tooltip placement="top-end" />
    </Space>
    <Space
      size={spaceSize}
      direction="vertical"
      style={{
        display: 'flex',
        width: buttonWidth,
        marginTop: spaceSize,
        float: 'left',
      }}
    >
      <Tooltip placement="left-start" />
      <Tooltip placement="left" />
      <Tooltip placement="left-end" />
    </Space>
    <Space
      size={spaceSize}
      direction="vertical"
      style={{
        display: 'flex',
        width: buttonWidth,
        marginTop: spaceSize,
        marginLeft: buttonWidth * 4 + spaceSize * 4,
      }}
    >
      <Tooltip placement="right-start" />
      <Tooltip placement="right" />
      <Tooltip placement="right-end" />
    </Space>
    <Space
      size={spaceSize}
      style={{
        display: 'flex',
        marginLeft: buttonWidth + spaceSize,
        marginTop: spaceSize,
        clear: 'both',
      }}
    >
      <Tooltip placement="bottom-start" />
      <Tooltip placement="bottom" />
      <Tooltip placement="bottom-end" />
    </Space>
    <div style={{ height: 1200 }} />
  </PopperBoundary>
);

// const ContextMenuEffect = React.forwardRef<VirtualElement, React.>(() => {

// });

// export const OnContextMenu = () => {

// };
// const CloneElementContext = React.createContext();
// const Child: React.FC<{ state: boolean }> = ({ children, state }) => {
//   useLogger('Child');
//   // const state_ = React.useContext(CloneElementContext);
//   const state_ = state;
//   return (
//     <>
//       {JSON.stringify(state_)}
//       <Button
//         state_
//         ref={React.useCallback(
//           value => console.log('refCallback 122', value),
//           []
//         )}
//       >
//         {'122'}
//         {JSON.stringify(state_)}
//       </Button>
//       {React.cloneElement(children, {
//         // ref: useProxyRef(null, value => console.log('useProxyRef', value)),
//         ref: React.useCallback(
//           value => console.log('refCallback 123', value),
//           []
//         ),
//         state_,
//       })}
//     </>
//   );
// };
// export const CloneElement = () => {
//   const [state, setState] = React.useState(false);
//   useLogger('CloneElement');
//   return (
//     <CloneElementContext.Provider value={state}>
//       <button onClick={() => setState(s => !s)}>{'forceUpdate'}</button>
//       <Child state={state}>
//         <Button>{'123'}</Button>
//       </Child>
//     </CloneElementContext.Provider>
//   );
// };

const ActionTestRecursizePopover = ({
  prefix = 'root',
  size = 2,
  ...restProps
}: {
  prefix?: string;
  size?: number;
} & Pick<PopperProps, 'trigger' | 'showAction' | 'hideAction' | 'disabled'>) => (
  <>
    {Array(size)
      .fill(0)
      .map((_, i) => (
        <Popper
          arrow={<PopperArrow />}
          offset={10}
          popupClassName={`${prefix}-${i}-popup`}
          key={i}
          popup={
            <div
              id={`${prefix}-${i}-popup`}
              style={{
                border: '1px solid var(--line-border-component)',
                background: 'white',
                padding: 16,
              }}
            >
              <ActionTestRecursizePopover prefix={`${prefix}-${i}`} {...restProps} />
            </div>
          }
          {...restProps}
        >
          <button
            id={`${prefix}-${i}-button`}
            type="button"
            style={{
              margin: 16,
            }}
          >
            {`${prefix}-${i}`}
          </button>
        </Popper>
      ))}
  </>
);

export const ActionTest = () => {
  const props: React.ComponentProps<typeof ActionTestRecursizePopover>[] = [
    {
      trigger: 'hover',
    },
    {
      trigger: 'click',
    },
    {
      trigger: 'contextMenu',
    },
    {
      trigger: 'hover',
      hideAction: ['click'],
    },
    {
      trigger: 'click',
      hideAction: ['mouseLeave'],
    },
    {
      trigger: 'hover',
      showAction: ['click'],
    },
    {
      trigger: 'click',
      showAction: ['mouseEnter'],
    },
    {
      showAction: ['click'],
      hideAction: ['mouseLeave'],
    },
    {
      showAction: ['contextMenu'],
      hideAction: ['mouseLeave'],
    },
    {
      showAction: ['mouseEnter'],
      hideAction: ['click'],
    },
  ];
  return (
    <>
      {props.map(o => (
        <div key={JSON.stringify(o)}>
          <h6>{JSON.stringify(o)}</h6>
          <ActionTestRecursizePopover {...o} />
        </div>
      ))}
    </>
  );
};

interface FormValues {
  adjustY: boolean;
  adjustX: boolean;
  open: boolean;
  placement: PositionType;
  offset: { x: number; y: number };
}

export const ComparePopupAlign = () => (
  <div>
    <div style={{ height: 800 }} />
    <Form<FormValues>
      layout="inline"
      initialValues={{
        adjustY: false,
        adjustX: false,
        open: true,
        placement: 'top',
        offset: {
          x: 8,
          y: -4,
        },
      }}
    >
      <Form.Item label="Open" name="open" valuePropName="checked">
        <Switch />
      </Form.Item>
      <Form.Item label="AdjustX" name="adjustX" valuePropName="checked">
        <Checkbox />
      </Form.Item>
      <Form.Item label="AdjustY" name="adjustY" valuePropName="checked">
        <Checkbox />
      </Form.Item>
      <Form.Item label="Placement" name="placement">
        <Select
          style={{ width: 200 }}
          options={PositionEnum.map(o => ({ value: o }))}
          allowClear={false}
          dropdownMatchSelectWidth={false}
        />
      </Form.Item>
      <Form.Item label="Offset.x" name={['offset', 'x']}>
        <InputNumber />
      </Form.Item>
      <Form.Item label="Offset.y" name={['offset', 'y']}>
        <InputNumber />
      </Form.Item>
      <FormSelector selector={(o: FormContextType<FormValues>) => o.values}>
        {({ adjustX, adjustY, open, placement, offset }) => {
          const align = {
            points: getPlacements({
              offset: 0,
              defaultOffset: 0,
            })[placement].points,
            offset: [offset.x, offset.y],
            overflow: {
              adjustX,
              adjustY,
            },
          };
          const content = Array(100)
            .fill(0)
            .map((_, i) => i)
            .join(', ');
          return (
            <Space direction="vertical" inline={false} align="center">
              <Space
                style={{
                  width: '100%',
                  justifyContent: 'center',
                  marginTop: 200,
                }}
                size={400}
              >
                <ConfigProvider>
                  <OldPopper visible={open} trigger="click" content={content} popupAlign={align} placement="bottomLeft">
                    <Button>Popover</Button>
                  </OldPopper>
                </ConfigProvider>
                <ConfigProvider PopperSidecar={FloatingUIPopperSidecar}>
                  <OldPopper visible={open} trigger="click" content={content} popupAlign={align} placement="bottomLeft">
                    <Button>Popover, FloatingUIPopperSidecar</Button>
                  </OldPopper>
                </ConfigProvider>
              </Space>
              <Space
                style={{
                  width: '100%',
                  justifyContent: 'center',
                  marginTop: 200,
                }}
                size={400}
              >
                <ConfigProvider>
                  <Dropdown
                    visible={open}
                    trigger="click"
                    overlay={<div className={popoverCss}>{content}</div>}
                    align={align}
                    placement="bottomLeft"
                  >
                    <Button>Dropdown</Button>
                  </Dropdown>
                </ConfigProvider>
                <ConfigProvider PopperSidecar={FloatingUIPopperSidecar}>
                  <Dropdown
                    visible={open}
                    trigger="click"
                    overlay={<div className={popoverCss}>{content}</div>}
                    align={align}
                    placement="bottomLeft"
                  >
                    <Button>Dropdown, FloatingUIPopperSidecar</Button>
                  </Dropdown>
                </ConfigProvider>
              </Space>
            </Space>
          );
        }}
      </FormSelector>
    </Form>
    <div style={{ height: 800 }} />
  </div>
);
