// @universe-design/react/src/shared/components/popper/PopperProvider.tsx
import * as React from 'react';
import { useControlledState } from '@universe-design/react-utilities';
import composeRefs from '../../../util/composeRefs';
import { usePopperShared } from './usePopper';
import type { UsePopperActionSharedProps } from './usePopperAction';
import { usePopperActionShared } from './usePopperAction';
import { deprecatedWarning, __SCREENSHOT__, warning } from '../../../util';
import { isOverflowed as defaultIsOverflow } from '../../../util/dom';
import { useCssScope } from '../../../config-provider';
import type { PopperContextType } from './PopperContext';
import { PopperContextProvider } from './PopperContext';

export interface PopperRef extends Pick<PopperContextType, 'visible' | 'setVisible'> {
  /**
   * 重新定位
   */
  forcePopupAlign: () => void;
}

export interface PopperProviderProps extends UsePopperActionSharedProps {
  /**
   * 弹层是否可见的受控值
   */
  visible?: boolean;
  /**
   * 弹层是否可见的默认值
   * @defaultValue false
   */
  defaultVisible?: boolean;
  /**
   * 弹层显示隐藏的回调
   */
  onVisibleChange?: (visible: boolean) => void;
  /**
   * @deprecated udTriggerRef 即将废弃，可以直接使用 ref 获取命令调用方法
   */
  udTriggerRef?: React.Ref<PopperRef>;
  /**
   * 禁用弹层组件，禁用所有可能损耗性能的初始化逻辑
   * @defaultValue false
   */
  disabled?: boolean;
  prefixCls?: string;
  /**
   * 修饰符，提供 trigger 的内建增强功能
   * @deprecated 已废弃，请使用 `eventListeners` 和 `maxHeight` 属性
   * @defaultValue `{ forcePopupAlign: true, popupOverflowAuto: true }`
   */
  modifiers?: {
    /**
     * 所有滚动父级的滚动事件、 window 的 resize 事件、trigger 和 popup resize 事件触发时强制进行 popup 的对齐
     */
    forcePopupAlign?: boolean;
    /**
     * popup 在不进行 adjust align 的时候可以内部滚动
     */
    popupOverflowAuto?: boolean;
  };
  /**
   * 是否开启溢出显示文字提示
   * @defaultValue false
   */
  checkOverflow?: boolean | ((ele: Element) => boolean);
  /**
   * screenshot 时是否始终打开 popper
   * @internal
   * @defaultValue false
   */
  __screenshotAlwaysOpenEnabled?: boolean;
}

export const PopperProvider = React.forwardRef<PopperRef, React.PropsWithChildren<PopperProviderProps>>(
  (props, ref) => {
    const {
      visible: visibleProp,
      checkOverflow = false,
      defaultVisible = false,
      children,
      trigger: triggerProp,
      showAction: showActionProp,
      hideAction: hideActionProp,
      mouseEnterDelay,
      mouseLeaveDelay: mouseLeaveDelayProp,
      onVisibleChange,
      udTriggerRef,
      disabled = false,
      prefixCls,
      modifiers,
      __screenshotAlwaysOpenEnabled = false,
    } = props;

    /**
     * Deprecated Warning
     */
    warning(!udTriggerRef, deprecatedWarning('Popper', 'udTriggerRef', 'ref'));
    warning(!modifiers, deprecatedWarning('Popper', 'modifiers', ['eventListeners', 'enableMaxHeight']));

    const rootPrefixCls = useCssScope('popper', prefixCls)('');

    /**
     * Popper 打开状态
     */
    const [visible, _setVisible] = useControlledState(defaultVisible, visibleProp, { onChange: onVisibleChange });

    /**
     * inline
     */
    const inlineElementMousePositionRef = React.useRef<{ x: number; y: number }>({
      x: 0,
      y: 0,
    });

    /**
     * Popper 对齐
     */
    const { referenceRef, referenceElementRef, referenceElement, popperLatestValues } = usePopperShared();

    const checkVisible = React.useCallback<React.Dispatch<boolean>>(
      _visible => {
        if (_visible && checkOverflow) {
          const overflowChecker = typeof checkOverflow === 'function' ? checkOverflow : defaultIsOverflow;
          const child = referenceElementRef.current;
          if (!child || !overflowChecker(child)) {
            _setVisible(false);
            return;
          }
        }
        _setVisible(_visible);
      },
      [checkOverflow, _setVisible, referenceElementRef],
    );

    // 保证在截图时一定能弹出
    __SCREENSHOT__ &&
      __screenshotAlwaysOpenEnabled &&
      // eslint-disable-next-line react-hooks/rules-of-hooks
      React.useLayoutEffect(() => {
        checkVisible(true);
      });

    /**
     * Popper 触发方式
     */
    const {
      handleReferenceClick,
      handleContextMenu,
      handleMouseEnter,
      handleMouseLeave,
      handleKeyDown,
      handleReferenceFocus,
      setVisible,
      showAction,
      hideAction,
      alignPoint,
      mouseLeaveDelay,
      setVisibleDelay,
    } = usePopperActionShared(
      {
        trigger: triggerProp,
        showAction: showActionProp,
        hideAction: hideActionProp,
        mouseEnterDelay,
        mouseLeaveDelay: mouseLeaveDelayProp,
      },
      {
        visible,
        setVisible: checkVisible,
        referenceRef,
        referenceElementRef,
        inlineElementMousePositionRef,
      },
    );

    const mergedPopperRef = React.useMemo(() => composeRefs(udTriggerRef, ref), [ref, udTriggerRef]);

    React.useImperativeHandle(
      mergedPopperRef,
      () => ({
        visible,
        setVisible,
        forcePopupAlign: () => popperLatestValues.update?.(),
      }),
      [popperLatestValues, setVisible, visible],
    );

    return (
      <PopperContextProvider
        value={{
          showAction,
          hideAction,
          alignPoint,
          visible,
          handleReferenceClick,
          handleContextMenu,
          handleMouseEnter,
          handleMouseLeave,
          handleKeyDown,
          handleReferenceFocus,
          setVisible,
          setVisibleDelay,
          referenceRef,
          referenceElement,
          referenceElementRef,
          popperLatestValues,
          disabled,
          rootPrefixCls,
          mouseLeaveDelay,
          inlineElementMousePositionRef,
        }}
      >
        {children}
      </PopperContextProvider>
    );
  },
);

if (__DEV__) {
  PopperProvider.displayName = 'PopperProvider';
}
