// @universe-design/react/src/shared/components/popper/PopperPopupVisibleController.tsx
import React from 'react';
import type { TransitionStatus } from 'react-transition-group';
import { Transition } from 'react-transition-group';
import type { UseAnimatedVisibleProps } from '../transition/useAnimatedVisible';
import { useTransitionGroupNoopNodeRef } from '../transition/useAnimatedVisible';
import { defaults } from './types';
import { isFunction } from '../../../util';
import { usePopperContext } from './PopperContext';

export interface PopperPopupVisibleControllerProps extends Omit<UseAnimatedVisibleProps, 'destroyOnClose' | 'visible'> {
  /**
   * 是否在关闭的时候卸载节点
   * @defaultValue true
   */
  destroyOnHide?: boolean;
  /**
   * 渐变的持续时间
   */
  timeout?: {
    appear?: number;
    enter?: number;
    exit?: number;
  };
}

export const PopperPopupVisibleController = (
  props: PopperPopupVisibleControllerProps & {
    children: (status: TransitionStatus) => React.ReactNode;
  },
) => {
  const {
    timeout = defaults.timeout,
    forceRender = false,
    destroyOnHide = true,
    onAfterClose,
    onAfterOpen,
    children,
  } = props;

  const { visible, popperLatestValues } = usePopperContext();

  const noopRef = useTransitionGroupNoopNodeRef();

  return (
    <Transition
      nodeRef={noopRef}
      timeout={timeout}
      appear
      {...{
        in: visible,
        mountOnEnter: !forceRender,
        unmountOnExit: destroyOnHide,
        onEnter: () => {
          popperLatestValues.update?.();
        },
        onEntered: () => {
          isFunction(onAfterOpen) && onAfterOpen();
        },
        onExited: () => {
          isFunction(onAfterClose) && onAfterClose();
          popperLatestValues.resetPlacement?.();
        },
      }}
    >
      {children}
    </Transition>
  );
};
