// @universe-design/react/src/shared/components/popper/PopperReference.tsx
import * as React from 'react';
import clsx from 'clsx';
import { useCloneElementRef } from '@universe-design/react-utilities';
import { usePopperContext } from './PopperContext';
import { isFunction, isNonNullable } from '../../../util';
import { composeEventHandlers } from '../../../util/function';
import { useEventCallback } from '../../../hooks';

export type PopperReferenceChildProps = Pick<
  React.HTMLAttributes<Element>,
  | 'onClick'
  | 'onMouseEnter'
  | 'onMouseLeave'
  | 'onContextMenu'
  | 'onKeyDown'
  | 'onFocus'
  | 'onBlur'
  | 'className'
  | 'style'
> & { 'aria-expanded'?: boolean };

export interface PopperReferenceProps extends PopperReferenceChildProps {
  children:
    | React.ReactElement<PopperReferenceChildProps>
    | ((props: { visible: boolean }) => React.ReactElement<PopperReferenceChildProps>);
  /**
   * 弹层打开时触发元素的 `className` 属性透传
   */
  openClassName?: string;
  /**
   * `className` 属性透传
   */
  className?: string;
  /**
   * `style` 属性透传
   */
  style?: React.CSSProperties;
  /**
   * @internal
   * reference ref 是否只接受 dom
   */
  domOnly?: boolean;
}

export const PopperReference = (props: PopperReferenceProps) => {
  const {
    rootPrefixCls,
    referenceRef,
    handleReferenceClick,
    handleMouseEnter,
    handleMouseLeave,
    handleContextMenu,
    handleKeyDown,
    handleReferenceFocus,
    alignPoint,
    disabled,
    visible,
  } = usePopperContext();

  const { children: childrenProp, className, style, openClassName = `${rootPrefixCls}-open`, domOnly = true } = props;

  // resolve function children
  let children = isFunction(childrenProp) ? childrenProp({ visible }) : childrenProp;

  const mergedClassName = clsx(children.props.className, className, visible && openClassName);

  const eventProps: PopperReferenceChildProps = {
    onClick: useEventCallback(composeEventHandlers(children.props.onClick, props.onClick, handleReferenceClick)),
    onMouseEnter: useEventCallback(
      composeEventHandlers(children.props.onMouseEnter, props.onMouseEnter, handleMouseEnter),
    ),
    onMouseLeave: useEventCallback(
      composeEventHandlers(children.props.onMouseLeave, props.onMouseLeave, handleMouseLeave),
    ),
    onFocus: useEventCallback(composeEventHandlers(children.props.onFocus, props.onFocus, handleReferenceFocus)),
    onContextMenu: useEventCallback(
      composeEventHandlers(children.props.onContextMenu, props.onContextMenu, handleContextMenu),
    ),
    onKeyDown: useEventCallback(composeEventHandlers(children.props.onKeyDown, props.onKeyDown, handleKeyDown)),
  };

  // clone props
  children = React.cloneElement(children, {
    ...(mergedClassName ? { className: mergedClassName } : {}),
    style: { ...style, ...children.props.style },
    ...(visible && !isNonNullable(children.props['aria-expanded']) ? { 'aria-expanded': visible } : {}),
    ...(disabled ? {} : eventProps),
  });

  // clone ref
  children = useCloneElementRef({ children, ref: referenceRef }, { skipped: alignPoint || disabled, domOnly });

  return children;
};
