// @universe-design/react/src/shared/components/popper/PopperPopup.tsx
import React from 'react';
import clsx from 'clsx';
import Portal from '../../../portal';
import { ZoomTransition } from '../transition/ZoomTransition';
import { StreamlineArrow } from '../trigger/StreamlineArrow';
import { PopperDrillingContextProvider } from './PopperDrillingContext';
import { FocusLock } from '../FocusLock';
import type { RefThunkElement, ThunkElement } from '../../../util/getThunkElement';
import type { FocusableElement } from '../../../util';
import { deprecatedWarning, isFunction, warning } from '../../../util';
import { PopperActionEnum, usePopperPopupAction } from './usePopperAction';
import type { UsePopperPopupProps } from './usePopper';
import { usePopperPopup } from './usePopper';
import { PopperPopupContextProvider, usePopperPopupContext } from './PopperPopupContext';
import { usePopperContext } from './PopperContext';
import { defaults } from './types';
import type { PopperPopupVisibleControllerProps } from './PopperPopupVisibleController';
import type { TransitionStatus } from 'react-transition-group';
import { useConstant } from '@universe-design/react-utilities';
import { usePopupPortalContainer } from '@universe-design/react-provider';
// import { useLogger } from '../../../hooks';

export const PopperArrow = ({ hasStroke }: { hasStroke?: boolean }) => {
  const { rootPrefixCls } = usePopperContext();
  const { placement } = usePopperPopupContext();

  const direction = (() => {
    if (!placement) {
      return undefined;
    }
    if (placement.startsWith('right') || placement.startsWith('left')) {
      return 'vertical';
    }
    if (placement.startsWith('top') || placement.startsWith('bottom')) {
      return 'horizontal';
    }
  })();

  return (
    <StreamlineArrow
      direction={direction}
      className={`${rootPrefixCls}__arrow__streamline`}
      hasStroke={hasStroke}
      style={{
        display: 'flex',
        transform: (() => {
          // 由于 arrow 尺寸跟 placement 有关系，所以不能占 arrow container 的宽度，不然定位的时候是 arrow 大小是不对的，定位以后 arrow 更新会歪掉，但是这样写就意味着 arrowPadding 要手动加上 arrow 一半的宽度
          if (!placement) {
            return;
          }
          if (placement.startsWith('right')) {
            return `translate(${hasStroke ? 'calc(1px - 100%)' : '-100%'}, -50%) rotate(180deg)`;
          }
          if (placement.startsWith('left')) {
            return `translate(${hasStroke ? '-1px' : '0'}, -50%)`;
          }
          if (placement.startsWith('top')) {
            return `translate(-50%, ${hasStroke ? '-1px' : '0'})`;
          }
          if (placement.startsWith('bottom')) {
            return `translate(-50%, ${hasStroke ? 'calc(1px - 100%)' : '-100%'}) rotate(180deg)`;
          }
        })(),
        zIndex: 'inherit',
        position: 'absolute',
      }}
    />
  );
};

export interface PopperPopupProps
  extends Omit<UsePopperPopupProps, 'inline'>,
    Pick<PopperPopupVisibleControllerProps, 'timeout'> {
  /**
   * 可供用户自由替换的 `Transition` 组件
   */
  Transition?: typeof ZoomTransition;
  /**
   * 箭头
   */
  arrow?: React.ReactNode;
  /**
   * 弹层元素的 `className` 属性透传
   */
  popupClassName?: string;
  /**
   * 弹层元素的 `style` 属性透传
   */
  popupStyle?: React.CSSProperties;
  /**
   * (PopperSidecar=[FloatingUIPopperSidecar](https://universe.bytedance.net/develop/react/config-provider#%E4%BD%BF%E7%94%A8%E5%8A%9F%E8%83%BD%E6%9B%B4%E5%BC%BA%E5%A4%A7%E7%9A%84-popper-%E7%BB%84%E4%BB%B6)) popup 挂载的父节点。
   *
   * 如果只是为了实现滚动跟随，我们不建议用户设置该属性，因为会产生层级问题，并且弹层无法超出滚动容器显示，例如 `Dialog` 弹层场景。`FloatingUIPopperSidecar` 弹层组件默认支持滚动跟随。
   *
   * **注意** 该方法仅在使用 `FloatingUIPopperSidecar` 时有效，旧版弹层请使用 `getPopupContainer`。
   */
  popupPortalContainer?: false | null | RefThunkElement<Element> | ((reference: Element | null) => Element | null);
  /**
   * @deprecated rc 的旧签名，不准确，准备废弃，建议开启 `FloatingUIPopperSidecar` 并使用 `popupPortalContainer` 属性。
   *
   * 目前 `getPopupContainer` 属性在使用 `RcTriggerPopperSidecar` 或 `FloatingUIPopperSidecar` 时均有效。
   *
   * @see [FloatingUIPopperSidecar](https://universe.bytedance.net/develop/react/config-provider#%E4%BD%BF%E7%94%A8%E5%8A%9F%E8%83%BD%E6%9B%B4%E5%BC%BA%E5%A4%A7%E7%9A%84-popper-%E7%BB%84%E4%BB%B6)
   */
  getPopupContainer?: (node: HTMLElement) => HTMLElement;
  /**
   * (PopperSidecar=[FloatingUIPopperSidecar](https://universe.bytedance.net/develop/react/config-provider#%E4%BD%BF%E7%94%A8%E5%8A%9F%E8%83%BD%E6%9B%B4%E5%BC%BA%E5%A4%A7%E7%9A%84-popper-%E7%BB%84%E4%BB%B6)) popup 打开时初始 focus 的元素, 默认为 popup content
   */
  initialFocus?: ThunkElement<FocusableElement>;
  /**
   * (PopperSidecar=[FloatingUIPopperSidecar](https://universe.bytedance.net/develop/react/config-provider#%E4%BD%BF%E7%94%A8%E5%8A%9F%E8%83%BD%E6%9B%B4%E5%BC%BA%E5%A4%A7%E7%9A%84-popper-%E7%BB%84%E4%BB%B6)) 开启🤔关闭 focusLock
   * @defaultValue false
   */
  enableFocusLock?: boolean;
  /**
   * (PopperSidecar=[FloatingUIPopperSidecar](https://universe.bytedance.net/develop/react/config-provider#%E4%BD%BF%E7%94%A8%E5%8A%9F%E8%83%BD%E6%9B%B4%E5%BC%BA%E5%A4%A7%E7%9A%84-popper-%E7%BB%84%E4%BB%B6)) 关闭对话框后是否恢复 focus 到 reference 元素
   * @defaultValue true
   */
  enableRestoreFocus?: boolean;
  /**
   * (PopperSidecar=[FloatingUIPopperSidecar](https://universe.bytedance.net/develop/react/config-provider#%E4%BD%BF%E7%94%A8%E5%8A%9F%E8%83%BD%E6%9B%B4%E5%BC%BA%E5%A4%A7%E7%9A%84-popper-%E7%BB%84%E4%BB%B6)) 优化跨行内联元素的定位。
   * @defaultValue false
   */
  inline?: boolean;
  // 控制是否阻止 popup 点击事件冒泡，设置 click 为 true 时阻止事件传播
  popupStopPropagation?: { click?: boolean };
}

export const PopperPopup: React.FC<
  PopperPopupProps & {
    /**
     * @internal
     * 用于 PopperPopupVisibleController 给 PopperPopup 传值
     */
    transitionStatus: TransitionStatus;
  }
> = props => {
  const {
    children,
    Transition: UserTransition = ZoomTransition,
    timeout = defaults.timeout,
    arrow,
    popupPortalContainer,
    getPopupContainer,
    popupClassName,
    popupStyle,
    initialFocus,
    enableFocusLock = false,
    enableRestoreFocus = true,
    inline = false,
    transitionStatus: status,
    popupStopPropagation = {},
    ...popperProps
  } = props;

  const { click: popupClickStopPropagation = false } = popupStopPropagation;

  /**
   * Deprecated Warning
   */
  warning(!getPopupContainer, deprecatedWarning('Popper', 'getPopupContainer', 'popupPortalContainer'));

  const {
    handleMouseEnter,
    handleMouseLeave,
    handleKeyDown,
    visible,
    referenceElementRef,
    rootPrefixCls,
    showAction,
    mouseLeaveDelay,
    disabled,
    setVisibleDelay,
    inlineElementMousePositionRef,
  } = usePopperContext();
  const { floatingRef, arrowRef, placement, getScrollParents, referenceHidden } = usePopperPopup(
    {
      ...popperProps,
      inline: inline
        ? {
            get x() {
              return inlineElementMousePositionRef.current.x;
            },
            get y() {
              return inlineElementMousePositionRef.current.y;
            },
          }
        : undefined,
    },
    {
      skipped: disabled || !visible,
    },
  );
  const { handlePopupFocus, handlePopperMouseDownCapture } = usePopperPopupAction(
    {},
    { referenceElementRef, floatingRef, visible, setVisibleDelay },
  );

  // Transition
  const animatedNodeRef = React.useRef<HTMLDivElement>(null);

  // FocusLock
  const shards = useConstant(() => [floatingRef, referenceElementRef]);

  // getPopupContainer
  const ctxGetPopupContainer = usePopupPortalContainer();
  const portalContainer = React.useMemo(() => {
    if (popupPortalContainer === false) {
      // non portal
      return false;
    }
    if (popupPortalContainer === null) {
      // default portal
      return null;
    }
    return () => {
      if (isFunction(popupPortalContainer)) {
        return popupPortalContainer(referenceElementRef.current);
      }
      if (popupPortalContainer) {
        return popupPortalContainer.current;
      }
      if (referenceElementRef.current) {
        return (getPopupContainer ?? ctxGetPopupContainer)(referenceElementRef.current as HTMLElement);
      }
      return null;
    };
  }, [ctxGetPopupContainer, getPopupContainer, popupPortalContainer, referenceElementRef]);

  const mergedInitialFocusRef = initialFocus === null ? null : initialFocus ?? animatedNodeRef;

  // 🐛
  // useLogger('PopperPopup', {
  //   placement,
  //   referenceHidden: drillingCtx?.referenceHidden,
  // });

  // 🐛
  // console.log('Transition', {
  //   status,
  //   placement,
  // });
  const animatedVisible = status !== 'exited';
  const isHidden = referenceHidden ?? false;
  const isNoEntry = mouseLeaveDelay === 0;

  const lockProps: React.ComponentPropsWithoutRef<'section'> & {
    'data-popper-animated': true;
  } = {
    'data-popper-animated': true,
    style: {
      display: 'flex',
      flexDirection: 'column',
      flex: 'auto',
      minWidth: 0,
      minHeight: 0,
      outline: 'none',
    },
    tabIndex: -1,
    role: 'dialog',
    onMouseEnter: handleMouseEnter,
    onMouseLeave: handleMouseLeave,
    onMouseDownCapture: handlePopperMouseDownCapture,
    onKeyDown: handleKeyDown,
    onClick: (e: React.MouseEvent) => {
      if (popupClickStopPropagation) {
        e.stopPropagation();
      }
    },
  };

  return (
    <PopperDrillingContextProvider
      value={{
        getScrollParents,
        referenceHidden,
      }}
    >
      <PopperPopupContextProvider value={{ placement }}>
        <Portal container={portalContainer}>
          <div
            className={clsx(rootPrefixCls, popupClassName)}
            style={{
              position: 'absolute',
              display: 'flex',
              flexDirection: 'column',
              ...(isHidden || !animatedVisible
                ? {
                    visibility: 'hidden',
                    opacity: 0,
                  }
                : {}),
              // 自动隐藏、无法移入以及 visible false（包括了 exit-active 阶段）的时候要让鼠标无法移入
              // TODO 需要补充测试。用 visible false 可以解决点击 popup 中取消按钮 close 过程中误触发 mouseEnter 导致再次 popup 再次打开
              ...(isHidden || isNoEntry || !visible ? { pointerEvents: 'none' } : {}),
              ...popupStyle,
            }}
            ref={floatingRef as unknown as React.RefCallback<HTMLElement>}
            {...{
              // 事件先挂在外层因为 FocusLock 的 lockProps 会覆盖 onBlur 和 onFocus https://github.com/theKashey/react-focus-lock/issues/189
              onFocus: handlePopupFocus,
            }}
          >
            <UserTransition
              nodeRef={animatedNodeRef}
              timeout={timeout}
              appear
              // 一定要在 positioning 状态第一次同步过后再做动画（因为首次进来的 placement 可能经过第一次 positioning 以后就变了），对于依赖 placement 的动画可能会不对，比如说 slide up down
              // https://bytedance.feishu.cn/wiki/wikcn5wrtUjqegriRR1zKpyzb3c
              in={visible && placement !== null}
              mountOnEnter={false}
              unmountOnExit={false}
              extraProps={{
                placement: placement ?? 'left', // 这里是随便给了一个 placement，由于我们只会在有 placement 的时候做动画，这个 fallback 只用作类型校验
              }}
            >
              {innerStatus => (
                //  A11y: When Popover opens, focus is sent to PopoverContent. When it closes, focus is returned to the reference. 这里一定要把定位元素和动画元素分开是因为他们的 transform 属性大概率会冲突
                <FocusLock
                  ref={animatedNodeRef}
                  initialFocusRef={mergedInitialFocusRef}
                  restoreFocus={
                    // hover 的时候关闭 restoreFocus，因为现在没有很好的办法解决 restoreFocus 到 reference 的时候导致 popup 再次打开的问题
                    !showAction.includes(PopperActionEnum.mouseEnter) && enableRestoreFocus
                  }
                  enableFocusLock={enableFocusLock}
                  shards={shards}
                  // 注意这里必须需要和 pointerEvents none 样式同步，当前元素是 pointerEvents none 时，focusLock 会 focus 到 body 或者 reference 导致一些问题。
                  // 不能用 animatedVisible 是因为关闭的时候就要立即取消，不能等待动画，不然就太慢了。
                  isDisabled={isHidden || !visible || isNoEntry}
                  as="section"
                  className={`${rootPrefixCls}__content`}
                  lockProps={
                    innerStatus === 'exited'
                      ? {
                          ...lockProps,
                          style: { ...lockProps.style, opacity: 0 },
                        }
                      : lockProps
                  }
                >
                  {children}
                  <div
                    className={`${rootPrefixCls}__arrow`}
                    style={{
                      position: 'absolute',
                      lineHeight: 0, // 不加这个在 safari 上高度会变大
                    }}
                    ref={arrowRef}
                  >
                    {arrow}
                  </div>
                </FocusLock>
              )}
            </UserTransition>
          </div>
        </Portal>
      </PopperPopupContextProvider>
    </PopperDrillingContextProvider>
  );
};
