// @universe-design/react/src/shared/components/popper/usePopper.ts
/* eslint-disable max-nested-callbacks */
/* eslint-disable @typescript-eslint/no-use-before-define */
import * as React from 'react';
import type { Middleware, VirtualElement, ComputePositionReturn, RootBoundary } from '@floating-ui/core';
import { offset, shift, flip, arrow, size, hide, inline } from '@floating-ui/core';
import type { DetectOverflowOptions } from '@floating-ui/dom';
import { computePosition, getOverflowAncestors } from '@floating-ui/dom';
import uniq from 'lodash/uniq';
import { BehaviorSubject, from, EMPTY } from 'rxjs';
import { switchMap } from 'rxjs/operators';
import type { Placement } from './types';
import type { HookSharedOptions } from '../../../util';
import { getWindow, isFunction, isNonNullable } from '../../../util';
import { useIsomorphicLayoutEffect as useLayoutEffect, useEventCallback } from '../../../hooks';
import {
  useProxyRef,
  resizeObservable,
  useConstant,
  useWillUnmount,
  isElement,
} from '@universe-design/react-utilities';
import { useImmer } from 'use-immer';
import { usePopperDrillingContext } from './PopperDrillingContext';
import { getElementFromRefOrVirtualElement } from '../../../hooks/useComplementEvent';
import { usePopperBoundaryContext } from './PopperBoundary';
import { usePopperContext } from './PopperContext';

type OffsetOptions = NonNullable<Parameters<typeof offset>[0]>;
type FlipOptions = NonNullable<Parameters<typeof flip>[0]>;
type ShiftOptions = NonNullable<Parameters<typeof shift>[0]>;
type InlineOptions = NonNullable<Parameters<typeof inline>[0]>;

export interface UsePopperPopupProps {
  /**
   * popper 相对于其 reference 的位置。
   * @defaultValue `bottom`
   */
  placement?: Placement;
  /**
   * (PopperSidecar=[FloatingUIPopperSidecar](https://universe.bytedance.net/develop/react/config-provider#%E4%BD%BF%E7%94%A8%E5%8A%9F%E8%83%BD%E6%9B%B4%E5%BC%BA%E5%A4%A7%E7%9A%84-popper-%E7%BB%84%E4%BB%B6)) popper 元素相对于 reference 元素中的主轴偏移量。
   * @defaultValue 0
   */
  offset?: OffsetOptions;
  /**
   * (PopperSidecar=[FloatingUIPopperSidecar](https://universe.bytedance.net/develop/react/config-provider#%E4%BD%BF%E7%94%A8%E5%8A%9F%E8%83%BD%E6%9B%B4%E5%BC%BA%E5%A4%A7%E7%9A%84-popper-%E7%BB%84%E4%BB%B6)) 如果提供，可以在提供的事件触发时对 popper 进行重新定位，`scroll`: 在 reference 元素的所有 `scrollParents` 滚动时重新定位，`resize`: 在 `window` 触发 `resize` 事件时重新定位，`resizeObserver`: 在 reference 元素或 `popup` 元素尺寸变化的时候进行重新定位；
   * @defaultValue `{ scroll: true, resize: true, resizeObserver: true }`
   */
  eventListeners?: {
    scroll?: boolean;
    resize?: boolean;
    resizeObserver?: boolean;
  };
  /**
   * (PopperSidecar=[FloatingUIPopperSidecar](https://universe.bytedance.net/develop/react/config-provider#%E4%BD%BF%E7%94%A8%E5%8A%9F%E8%83%BD%E6%9B%B4%E5%BC%BA%E5%A4%A7%E7%9A%84-popper-%E7%BB%84%E4%BB%B6)) 如果为 true，将防止 popper 被超出截断，可以确保它在边界区域内可见。
   * @defaultValue true
   */
  shift?: boolean | Partial<ShiftOptions>;
  /**
   * (PopperSidecar=[FloatingUIPopperSidecar](https://universe.bytedance.net/develop/react/config-provider#%E4%BD%BF%E7%94%A8%E5%8A%9F%E8%83%BD%E6%9B%B4%E5%BC%BA%E5%A4%A7%E7%9A%84-popper-%E7%BB%84%E4%BB%B6)) popper 的根边界区域。
   * @defaultValue `viewport`
   */
  rootBoundary?: RootBoundary;
  /**
   * (PopperSidecar=[FloatingUIPopperSidecar](https://universe.bytedance.net/develop/react/config-provider#%E4%BD%BF%E7%94%A8%E5%8A%9F%E8%83%BD%E6%9B%B4%E5%BC%BA%E5%A4%A7%E7%9A%84-popper-%E7%BB%84%E4%BB%B6)) popper 的边界区域。在 `shift`、`flip`、`maxSize` 中间件中使用。
   * @defaultValue `clippingAncestors`
   */
  boundary?: DetectOverflowOptions['boundary'];
  /**
   * (PopperSidecar=[FloatingUIPopperSidecar](https://universe.bytedance.net/develop/react/config-provider#%E4%BD%BF%E7%94%A8%E5%8A%9F%E8%83%BD%E6%9B%B4%E5%BC%BA%E5%A4%A7%E7%9A%84-popper-%E7%BB%84%E4%BB%B6)) popper 距离边界的安全边距。
   * @defaultValue 16
   */
  boundaryPadding?: number;
  /**
   * (PopperSidecar=[FloatingUIPopperSidecar](https://universe.bytedance.net/develop/react/config-provider#%E4%BD%BF%E7%94%A8%E5%8A%9F%E8%83%BD%E6%9B%B4%E5%BC%BA%E5%A4%A7%E7%9A%84-popper-%E7%BB%84%E4%BB%B6)) 防止 arrow 出现在 popper 边缘的安全距离，与 popup 的圆角尺寸有关。
   * @defaultValue 0
   */
  arrowPadding?: number;
  /**
   * (PopperSidecar=[FloatingUIPopperSidecar](https://universe.bytedance.net/develop/react/config-provider#%E4%BD%BF%E7%94%A8%E5%8A%9F%E8%83%BD%E6%9B%B4%E5%BC%BA%E5%A4%A7%E7%9A%84-popper-%E7%BB%84%E4%BB%B6)) floating-ui 中间件数组。可以查看以下文档编写自己的中间件 [floating-ui middleware](https://floating-ui.com/docs/middleware)
   */
  middlewares?: Array<Middleware>;
  /**
   * (PopperSidecar=[FloatingUIPopperSidecar](https://universe.bytedance.net/develop/react/config-provider#%E4%BD%BF%E7%94%A8%E5%8A%9F%E8%83%BD%E6%9B%B4%E5%BC%BA%E5%A4%A7%E7%9A%84-popper-%E7%BB%84%E4%BB%B6)) 如果是 true，popper 将始终匹配 reference 的宽度。在上下弹出的组件，例如 `select` 几兄弟中使用十分有效。我们默认只会对上下弹出的情况进行宽度匹配，左右弹出的时候不开启。
   * @defaultValue false
   */
  matchReferenceWidth?: boolean;
  /**
   * (PopperSidecar=[FloatingUIPopperSidecar](https://universe.bytedance.net/develop/react/config-provider#%E4%BD%BF%E7%94%A8%E5%8A%9F%E8%83%BD%E6%9B%B4%E5%BC%BA%E5%A4%A7%E7%9A%84-popper-%E7%BB%84%E4%BB%B6)) popper 是否在原始位置放不下的时候，变化自己的位置。
   * @defaultValue true
   */
  flip?: boolean | Partial<FlipOptions>;
  /**
   * (PopperSidecar=[FloatingUIPopperSidecar](https://universe.bytedance.net/develop/react/config-provider#%E4%BD%BF%E7%94%A8%E5%8A%9F%E8%83%BD%E6%9B%B4%E5%BC%BA%E5%A4%A7%E7%9A%84-popper-%E7%BB%84%E4%BB%B6)) popper 是否根据边界设置最大高度，可以用于一些长列表场景下让 popper 做内部滚动。
   * @defaultValue true
   */
  enableMaxHeight?: boolean;
  /**
   * (PopperSidecar=[FloatingUIPopperSidecar](https://universe.bytedance.net/develop/react/config-provider#%E4%BD%BF%E7%94%A8%E5%8A%9F%E8%83%BD%E6%9B%B4%E5%BC%BA%E5%A4%A7%E7%9A%84-popper-%E7%BB%84%E4%BB%B6)) 当reference不可见时，隐藏（非卸载）popup；再次可见时，显示popup
   * @defaultValue true
   */
  enableAutoHide?: boolean;
  /**
   * (PopperSidecar=[FloatingUIPopperSidecar](https://universe.bytedance.net/develop/react/config-provider#%E4%BD%BF%E7%94%A8%E5%8A%9F%E8%83%BD%E6%9B%B4%E5%BC%BA%E5%A4%A7%E7%9A%84-popper-%E7%BB%84%E4%BB%B6)) 优化跨行内联元素的定位。
   */
  inline?: Partial<InlineOptions>;
}

const defaults = {
  middleware: [],
};

interface PopperLatestValues {
  updateByReference?: (reference: Element | VirtualElement | null | undefined) => void;
  update?: () => void;
  resetPlacement?: () => void;
}

/**
 * 这个 hook 的逻辑会在 invisible first render 运行，需要尽可能精简
 */
export function usePopperShared() {
  /**
   * ================================
   *
   * 从 Popup 注册到全局的方法
   *
   * ================================
   */
  const popperLatestValues = useConstant<PopperLatestValues>(() => ({
    updateByReference: undefined,
    update: undefined,
    resetPlacement: undefined,
  }));

  /**
   * ================================
   *
   * Reference 相关的 ref 方法
   *
   * ================================
   */
  const [referenceElement, setReferenceElement] = React.useState<Element | null>(null); // 用 state 存 reference element 的原因是 reference 的 element 变化很可能是子元素自己 rerender 导致的，而非父级 rerender。通过在 refCall 中调用 setState 让子元素的 element 变化父级可感知，这样父就可以对子进行事件的重新订阅。

  // 用于获取真实的 element
  const referenceElementRef = React.useRef<Element | null>(null);

  const referenceRef = useProxyRef<Element | VirtualElement>(null, function referenceRefCallback(reference) {
    // 🐛
    // console.log('PopperReference ref', reference);
    if (reference) {
      // 保证 referenceElementRef 里面一定不会存 null 值，这样 getPopupContainer 才能保证一定取到 reference element
      // **注意** 由于这里存的是上一次 rerender 后的非空值，本次 render 可能会变化造成问题
      referenceElementRef.current = getElementFromRefOrVirtualElement(reference);
    } else {
      referenceElementRef.current = null;
    }
    setReferenceElement(reference ? referenceElementRef.current : null);
    popperLatestValues.updateByReference?.(reference);
  });

  return {
    referenceRef,
    referenceElementRef,
    referenceElement,
    popperLatestValues,
  };
}

export type UsePopperSharedReturnType = ReturnType<typeof usePopperShared>;

/**
 * 注意这个 hook 的实现要尽可能 delay，避免在没有渲染的时候进行不必要的计算
 */
export function usePopperPopup(
  props: UsePopperPopupProps & {
    /**
     * @internal
     * 定位计算完成后的回调函数
     */
    onPositioned?: (computeData: ComputePositionReturn) => void;
  },
  { skipped = false }: HookSharedOptions = {},
) {
  const ctxBoundary = usePopperBoundaryContext();
  const {
    placement: placementProp = 'bottom',
    offset: offsetProp = 0,
    eventListeners = {},
    shift: shiftProp = true,
    rootBoundary = 'viewport',
    boundary = ctxBoundary ?? 'clippingAncestors',
    boundaryPadding = 16,
    arrowPadding = 0,
    middlewares: middlewaresProp = defaults.middleware,
    matchReferenceWidth = false,
    flip: flipProp = true,
    enableMaxHeight: maxHeight = true,
    enableAutoHide: autoHide = true,
    onPositioned,
    inline: inlineProp,
  } = props;
  const { scroll = true, resize = true, resizeObserver = true } = eventListeners;
  const { referenceElement, referenceRef, referenceElementRef, popperLatestValues } = usePopperContext();

  /**
   * Refs
   */
  type ScrollParents = (Element | Window | VisualViewport)[];
  const onPositionedLatest = useEventCallback(onPositioned);

  /**
   * Scroll & Resize Listener
   */
  const drillingCtx = usePopperDrillingContext();
  const getOwnScrollParents = useEventCallback((): ScrollParents => {
    const nodes = drillingCtx?.getScrollParents() ?? [];

    const elements = [referenceElementRef.current, floatingRefBeforeProxyRefCallback.current].filter(isElement);

    if (!elements.length) {
      return nodes;
    }

    const scrollParents = uniq(elements.map(el => getOverflowAncestors(el)).flat());

    return uniq([...scrollParents, ...nodes]);
  });

  /**
   * Floating resize
   */
  const floatingResizeObservableDisposer = React.useRef<(() => void) | null>(null);
  const unsubscribeToFloatingResize = () => {
    if (floatingResizeObservableDisposer.current) {
      floatingResizeObservableDisposer.current();
    }
  };
  const subscribeToFloatingResize = () => {
    const floating = floatingRefBeforeProxyRefCallback.current;
    if (skipped || !resizeObserver) {
      return;
    }
    if (floating) {
      floatingResizeObservableDisposer.current = resizeObservable.subscribe(floating, updateLatest);
    }
  };

  /**
   * Scoll
   */
  const floatingScrollParentListenersDisposer = React.useRef<(() => void) | null>(null);
  const unsubscribeToScroll = () => {
    if (floatingScrollParentListenersDisposer.current) {
      floatingScrollParentListenersDisposer.current();
    }
  };
  const subscribeToScroll = () => {
    if (skipped || !scroll) {
      return;
    }

    const nodes = getOwnScrollParents();

    if (!nodes.length) {
      return;
    }

    nodes.forEach(node => {
      node.addEventListener('scroll', updateLatest);
    });

    floatingScrollParentListenersDisposer.current = () => {
      nodes.forEach(node => {
        node.removeEventListener('scroll', updateLatest);
      });
    };
  };

  /**
   * Unsubscribe listeners during ref phase
   */
  useWillUnmount(() => {
    unsubscribeToFloatingResize();
    unsubscribeToScroll();
  });

  // 用于给 useProxyRef callback 中调用的函数获取最新的 floating
  const floatingRefBeforeProxyRefCallback = React.useRef<HTMLElement | null>(null);
  const floatingRef = useProxyRef<HTMLElement>(null, function floatingRefCallback(floating) {
    floatingRefBeforeProxyRefCallback.current = floating;
    // 🐛
    // console.log('PopperPopup ref', floating);
    _updateLatest({
      reference: referenceRef.current,
      floating,
    });

    // Floating resize
    unsubscribeToFloatingResize();
    subscribeToFloatingResize();

    // Listen to scroll
    unsubscribeToScroll();
    subscribeToScroll();
  });

  const arrowRef = React.useRef<HTMLDivElement>(null);

  /**
   * State
   */
  const [data, updateData] = useImmer<{
    placement: Placement | null;
    referenceHidden: boolean;
  }>(() => ({
    placement: null,
    referenceHidden: false,
  }));

  const middleware = React.useMemo<(Middleware | undefined)[]>(() => {
    /**
     * overflow options
     */
    const overflowOptions: Partial<DetectOverflowOptions> = {
      boundary,
      padding: boundaryPadding,
      rootBoundary,
    };

    /**
     * Options
     */
    const flipOptions: Partial<FlipOptions & DetectOverflowOptions> = (() => {
      if (!flipProp) {
        return {};
      }
      return {
        fallbackStrategy: 'bestFit',
        ...overflowOptions,
        ...(flipProp === true ? {} : flipProp),
      };
    })();

    const shiftOptions: Partial<ShiftOptions & DetectOverflowOptions> = (() => {
      if (!shiftProp) {
        return {};
      }
      return {
        ...overflowOptions,
        crossAxis: true, // 默认开启 crossAxis 为了实现当页面中 flip 怎么都放不下的时候盖住 reference
        ...(shiftProp === true ? {} : shiftProp),
      };
    })();

    return [
      inlineProp ? inline(inlineProp) : undefined,
      // {
      //   name: 'log',
      //   fn({ elements: { reference, floating }, rects }) {
      //     console.log(
      //       reference === referenceRef.current,
      //       floating === floatingRef.current,
      //       'inline',
      //       JSON.stringify(inlineProp),
      //       'rects.reference',
      //       rects.reference,
      //       'rects.floating',
      //       rects.floating,
      //       'placement',
      //       { placementProp, dataPlacement: data.placement }
      //     );
      //     return {};
      //   },
      // },
      matchReferenceWidth
        ? {
            name: 'matchReferenceWidth',
            fn({ placement: _placement, rects, elements: { floating } }) {
              if (['top', 'bottom'].includes(_placement.split('-')[0])) {
                rects.floating.width = rects.reference.width;
                (floating as HTMLElement).style.width = `${rects.reference.width}px`;
              }
              return {};
            },
          }
        : undefined,
      offset(offsetProp),
      autoHide
        ? hide({
            ...overflowOptions,
            padding: 0, // hide 逻辑不考虑安全边距，这样可以让 popup 在边缘时保留
          })
        : undefined,
      flipProp ? flip(flipOptions) : undefined,
      shiftProp ? shift(shiftOptions) : undefined,
      maxHeight && !inlineProp // TODO 暂时在 inline 开启的关闭 maxHeight https://github.com/floating-ui/floating-ui/issues/1570
        ? size({
            ...overflowOptions,
            apply({ availableHeight }) {
              const floatingEl = floatingRef.current;
              if (!floatingEl) {
                return;
              }
              Object.assign(floatingEl.style, {
                maxHeight: `${availableHeight}px`,
              });
            },
          })
        : undefined,
      {
        name: 'arrow',
        fn(args) {
          if (arrowRef.current !== null) {
            return arrow({
              element: arrowRef.current,
              padding: arrowPadding,
            }).fn(args);
          }
          return {};
        },
      },
      {
        name: 'transformOrigin',
        fn: ({ placement: _placement, rects, middlewareData: { arrow: _arrow } }) => {
          if (!_arrow) {
            return {};
          }
          const { x: arrowX, y: arrowY } = _arrow;
          const { width: popperW, height: popperH } = rects.floating;
          let transformOrigin = '';

          if (_placement.startsWith('right')) {
            transformOrigin = `0 ${((arrowY ?? 0) / popperH) * 100}%`;
          }
          if (_placement.startsWith('left')) {
            transformOrigin = `100% ${((arrowY ?? 0) / popperH) * 100}%`;
          }
          if (_placement.startsWith('top')) {
            transformOrigin = `${((arrowX ?? 0) / popperW) * 100}% 100%`;
          }
          if (_placement.startsWith('bottom')) {
            transformOrigin = `${((arrowX ?? 0) / popperW) * 100}% 0`;
          }

          return {
            data: {
              transformOrigin,
            },
          };
        },
      },
    ];
  }, [
    arrowPadding,
    autoHide,
    boundary,
    boundaryPadding,
    flipProp,
    floatingRef,
    inlineProp,
    matchReferenceWidth,
    maxHeight,
    offsetProp,
    rootBoundary,
    shiftProp,
  ]);

  const inputPopperElements$ = useConstant(
    () =>
      new BehaviorSubject<{
        reference: Element | VirtualElement | null | undefined;
        floating: HTMLElement | null;
        placement: Placement;
        middleware: Middleware[];
      }>({
        reference: null,
        floating: null,
        placement: placementProp,
        middleware: [],
      }),
  );

  const computePosition$ = useConstant(() =>
    inputPopperElements$.pipe(
      switchMap(({ reference, floating, placement, middleware: _middleware }) => {
        if (!reference || !floating) {
          return EMPTY;
        }
        return from(
          computePosition(reference, floating, {
            middleware: _middleware,
            placement,
            strategy: 'absolute',
          }).then(o => ({ ...o, floating })),
        );
      }),
    ),
  );

  React.useEffect(() => {
    const subscription = computePosition$.subscribe({
      next(computeData) {
        // 这里直接操作 dom 元素的 style 是因为这样可以不用等 react 调度，但是要保证 jsx 那边不要写 style，否则两边同名属性会互相覆盖
        const {
          x,
          y,
          placement,
          middlewareData: {
            hide: { referenceHidden = false } = {},
            arrow: { x: arrowX, y: arrowY } = {},
            transformOrigin,
          },
          floating,
        } = computeData;
        // 🐛
        // console.log('beforePositionApplied', computeData);
        updateData(draft => {
          draft.placement = computeData.placement;
          draft.referenceHidden = referenceHidden;
        });
        /**
         * Apply Floating Style
         */
        (function applyFloatingStyle() {
          floating.dataset.popperPlacement = placement;
          floating.style.top = '0px';
          floating.style.left = '0px';
          floating.style.transform = `translate3d(${Math.round(x)}px, ${Math.round(y)}px, 0)`;
        })();
        /**
         * Apply Arrow Style
         */
        (function applyArrowStyle() {
          const arrowEl = arrowRef.current;
          if (!arrowEl) {
            return;
          }
          // cleanup
          arrowEl.style.top = '';
          arrowEl.style.left = '';
          arrowEl.style.bottom = '';
          arrowEl.style.right = '';
          // eslint-disable-next-line max-nested-callbacks
          const staticSide = (() => {
            switch (placement.split('-')[0]) {
              case 'top':
                return 'bottom';
              case 'right':
                return 'left';
              case 'bottom':
                return 'top';
              case 'left':
                return 'right';
              default:
                throw new Error('Placement main side is not valid.');
            }
          })();
          arrowEl.style.left = isNonNullable(arrowX) ? `${arrowX}px` : '';
          arrowEl.style.top = isNonNullable(arrowY) ? `${arrowY}px` : '';
          arrowEl.style[staticSide] = '0px';
        })();
        /**
         * Apply Animated Node Style
         */
        (function applyAnimatedNodeStyle() {
          const animatedElement = floating.querySelector<HTMLElement>('[data-popper-animated]');
          if (!animatedElement) {
            return;
          }
          Object.assign(animatedElement.style, transformOrigin);
        })();
        isFunction(onPositionedLatest) && onPositionedLatest(computeData);
      },
      error(error) {
        console.error(error);
      },
    });
    return () => {
      subscription.unsubscribe();
    };
    // 这些都是不会变的值，如果有变化的值是不能放进来的，我们不希望多次订阅
  }, [computePosition$, onPositionedLatest, updateData]);

  const _update = React.useCallback(
    ({
      reference,
      floating,
    }: {
      reference: Element | VirtualElement | null | undefined;
      floating: HTMLElement | null;
    }) => {
      inputPopperElements$.next({
        reference,
        floating,
        placement: placementProp,
        middleware: [...middleware, ...middlewaresProp].filter(isNonNullable),
      });
    },
    [inputPopperElements$, placementProp, middleware, middlewaresProp],
  );
  const _updateLatest = useEventCallback(_update);

  const update = React.useCallback(() => {
    _update({
      reference: referenceRef.current,
      floating: floatingRef.current,
    });
  }, [_update, floatingRef, referenceRef]);
  const updateLatest = useEventCallback(update);

  /**
   * 在 popup 出现的时候往 popper 注入一些方法，这个写法主要是为了首屏尽可能少运行逻辑代码，尽可能把 visible 以后用到的逻辑都收敛到 popup 中
   */
  popperLatestValues.update = updateLatest;
  popperLatestValues.updateByReference = useEventCallback(reference => {
    _update({
      reference,
      floating: floatingRef.current,
    });
  });
  popperLatestValues.resetPlacement = React.useCallback(() => {
    updateData(draft => {
      draft.placement = null;
    });
  }, [updateData]);

  useWillUnmount(() => {
    popperLatestValues.resetPlacement = undefined;
    popperLatestValues.update = undefined;
    popperLatestValues.updateByReference = undefined;
  });

  /**
   * Window resize
   */
  React.useEffect(() => {
    if (skipped || !resize) {
      return;
    }

    const win = getWindow(referenceElement);

    win.addEventListener('resize', updateLatest);

    return () => {
      win.removeEventListener('resize', updateLatest);
    };
  }, [referenceElement, resize, skipped, updateLatest]);

  /**
   * Reference resize
   */
  React.useEffect(() => {
    // 保证在 skipped 变化时能重新订阅
    if (skipped || !resizeObserver) {
      return;
    }
    if (!referenceElement) {
      return;
    }
    return resizeObservable.subscribe(referenceElement, updateLatest);
  }, [referenceElement, resizeObserver, skipped, updateLatest]);

  useLayoutEffect(update, [update]);

  return {
    placement: data.placement,
    referenceHidden: drillingCtx?.referenceHidden || data.referenceHidden,
    floatingRef,
    arrowRef,
    getScrollParents: getOwnScrollParents,
  };
}

export type UsePopperPopupReturnType = ReturnType<typeof usePopperPopup>;
