// @universe-design/react/lib/date-picker/panels/selector/MonthYearSelector.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MonthYearSelector = void 0;
var _react = _interopRequireDefault(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _DownBoldOutlined = _interopRequireDefault(require("@universe-design/icons-react/DownBoldOutlined"));
var _hooks = require("../../../config-provider/hooks");
var _stylex2 = require("../../../stylex");
var _button = require("../../../button");
var _utils = require("../base/utils");
const MonthYearSelector = props => {
  const {
    className,
    mode,
    collapsed,
    pickerValue,
    hiddenCollapse,
    onYearClick,
    onMonthClick,
    onCollapsed
  } = props;
  const {
    locale: localeName,
    components: {
      datePicker: locale
    }
  } = (0, _hooks.useLocale)();
  const cssScope = (0, _hooks.useCssScope)('picker');
  const headerPrefixCls = cssScope('-panel-header');
  const wrapYear = children => /*#__PURE__*/_react.default.createElement("span", {
    key: "year",
    className: `${headerPrefixCls}-btn`,
    onClick: onYearClick
  }, children);
  const timeNodes = locale.monthYearSelector.f({
    year: yearFormat => wrapYear(pickerValue.clone().locale(localeName).format(yearFormat.toString())),
    month: monthFormat => /*#__PURE__*/_react.default.createElement("span", {
      key: "month",
      onClick: onMonthClick,
      className: `${headerPrefixCls}-btn`
    }, pickerValue.clone().locale(localeName).format(monthFormat.toString()))
  });
  return /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(`${headerPrefixCls}-content`, className, (0, _stylex.default)(_stylex2.cssTypographyHeadline))
  }, mode === 'year' ? wrapYear((0, _utils.getYearRange)(pickerValue)) : timeNodes, !hiddenCollapse && /*#__PURE__*/_react.default.createElement(_button.Button, {
    type: "icon",
    className: (0, _clsx.default)(`${headerPrefixCls}-icon`, `${headerPrefixCls}-collapse`),
    onClick: e => {
      e.stopPropagation();
      e.preventDefault();
      onCollapsed === null || onCollapsed === void 0 ? void 0 : onCollapsed(!collapsed);
    },
    size: "extra-small",
    icon: /*#__PURE__*/_react.default.createElement(_DownBoldOutlined.default, {
      className: (0, _clsx.default)({
        [`${headerPrefixCls}-rotate`]: true,
        [`${headerPrefixCls}-rotate180`]: !collapsed
      })
    })
  }));
};
exports.MonthYearSelector = MonthYearSelector;
