// @universe-design/react/src/carousel/stories/homeless.stories.tsx
import stylex from '@universe-design/stylex';
import * as React from 'react';
import { Carousel, Slider, Divider, CarouselSlide } from '../..';
import { forwardHTMLRef } from '../../util/forwardRef';
import type { CarouselProps } from '../Carousel';

export default {
  title: 'Components/Carousel/临时示例', // 注意：这里需要修改成自己对应 Component name。
  component: Carousel,
};

const urls = [
  'https://randomwordgenerator.com/img/picture-generator/57e5d2424a54a414f1dc8460962e33791c3ad6e04e507440762e7adc904cc6_640.jpg',
  'https://randomwordgenerator.com/img/picture-generator/52e3d14a4b5ab10ff3d8992cc12c30771037dbf85254784e77287cd1914a_640.jpg',
  'https://randomwordgenerator.com/img/picture-generator/53e0dd404350a914f1dc8460962e33791c3ad6e04e5074417c2d78d1904fc3_640.jpg',

  'https://randomwordgenerator.com/img/picture-generator/57e3d5444f53ac14f1dc8460962e33791c3ad6e04e507440772872d3914ccc_640.jpg',
  'https://randomwordgenerator.com/img/picture-generator/57e5d5434d55ad14f1dc8460962e33791c3ad6e04e507440722d72d1914ccc_640.jpg',
  'https://randomwordgenerator.com/img/picture-generator/57e8d2414c5bab14f1dc8460962e33791c3ad6e04e50744077297bd5964acd_640.jpg',
  'https://randomwordgenerator.com/img/picture-generator/51e3d4424d53b10ff3d8992cc12c30771037dbf85254784b7d297ed7934b_640.jpg',
  'https://randomwordgenerator.com/img/picture-generator/valentin-salja-vh7RmF-NjMc-unsplash.jpg',
  'https://randomwordgenerator.com/img/picture-generator/52e3dd404354b10ff3d8992cc12c30771037dbf852547940772c7cd59649_640.jpg',
  'https://randomwordgenerator.com/img/picture-generator/55e2dc434d52a914f1dc8460962e33791c3ad6e04e50744172277fd7914cc7_640.jpg',
  'https://randomwordgenerator.com/img/picture-generator/57e0d1424e50ac14f1dc8460962e33791c3ad6e04e507441722872d79644c7_640.jpg',
  'https://randomwordgenerator.com/img/picture-generator/54e7d14b4951af14f1dc8460962e33791c3ad6e04e50744172287ad3944ec7_640.jpg',
  'https://randomwordgenerator.com/img/picture-generator/55e2d3444c55a514f1dc8460962e33791c3ad6e04e50744172297cdd9e4ac2_640.jpg',
  'https://randomwordgenerator.com/img/picture-generator/55e0d1414d57ad14f1dc8460962e33791c3ad6e04e5074417c2d78d09045c4_640.jpg',
  'https://randomwordgenerator.com/img/picture-generator/54e2d1404a57a814f1dc8460962e33791c3ad6e04e5074417c2d78d19f44c4_640.jpg',
  'https://randomwordgenerator.com/img/picture-generator/57e9d2454f51a414f1dc8460962e33791c3ad6e04e507441722872d69f4bc6_640.jpg',
  'https://randomwordgenerator.com/img/picture-generator/54e6d4424c55ab14f1dc8460962e33791c3ad6e04e5074417d2e72d6914ac3_640.jpg',
  'https://randomwordgenerator.com/img/picture-generator/55e8dd424b51af14f1dc8460962e33791c3ad6e04e50744172287edd9745c5_640.jpg',
  'https://randomwordgenerator.com/img/picture-generator/55e2d7454251aa14f1dc8460962e33791c3ad6e04e507440772d73d69444c3_640.jpg',
  'https://randomwordgenerator.com/img/picture-generator/57e2d54a4c56ab14f1dc8460962e33791c3ad6e04e5074417c2e7dd29e4dc7_640.jpg',
];

const Box = forwardHTMLRef<{ width: number }, 'div'>(({ width }, ref) => (
  <div
    ref={ref}
    style={{ width }}
    className={stylex(stylex.create({ padding: 4, boxSizing: 'border-box', height: '100%' }))}
  >
    <div
      className={stylex(
        stylex.create({
          backgroundColor: $token['primary-content-default'],
          height: '100%',
        }),
      )}
    />
  </div>
));

export const AllCases = () => {
  const count = 36;
  const [slideIndex, setSlideIndex] = React.useState(0);
  return (
    <div style={{ padding: 36 }}>
      <Divider>Dynamic width</Divider>
      <Carousel style={{ height: 320 }}>
        <CarouselSlide fullWidth={false}>
          <Box width={200} />
        </CarouselSlide>
        <CarouselSlide fullWidth={false}>
          <Box width={80} />
        </CarouselSlide>
      </Carousel>
      <Divider>Pressure test: 1000</Divider>
      <Carousel style={{ height: 320 }}>
        {Array(1000)
          .fill(0)
          .map((_, i) => `https://www.dmoe.cc/random.php?id=${i}`)
          .map((src, i) => (
            <CarouselSlide
              key={i}
              fullWidth={false}
              style={{
                minWidth: 200,
              }}
            >
              <img src={src} style={{ height: '100%' }} />
            </CarouselSlide>
          ))}
      </Carousel>
      <Divider>Different indicator type and theme, controlled mode</Divider>
      <div style={{ padding: '0 36px' }}>
        <Slider value={slideIndex} min={0} max={count} onChange={setSlideIndex} step={1} />
      </div>
      {(() => {
        const cases: CarouselProps[] = [
          { indicatorProps: { type: 'dot', theme: 'light' } },
          { indicatorProps: { type: 'line', theme: 'light' } },
          {
            indicatorProps: { type: 'dot', theme: 'primary' },
            indicatorPosition: 'outer-bottom',
          },
          {
            indicatorProps: { type: 'line', theme: 'primary' },
            indicatorPosition: 'outer-bottom',
          },
        ];
        return cases.map(props => (
          <React.Fragment key={JSON.stringify(props)}>
            <Divider />
            <Carousel style={{ height: 320 }} slideIndex={slideIndex} onSlideIndexChange={setSlideIndex} {...props}>
              {Array(count)
                .fill(0)
                .map((_, i) => urls[i % urls.length])
                .map((src, i) => (
                  <CarouselSlide
                    key={i}
                    fullWidth={false}
                    style={{
                      minWidth: 200,
                    }}
                  >
                    <img
                      src={src}
                      alt={src}
                      style={{
                        height: '100%',
                      }}
                    />
                  </CarouselSlide>
                ))}
            </Carousel>
            <Divider />
            <Carousel style={{ height: 320 }} slideIndex={slideIndex} onSlideIndexChange={setSlideIndex} {...props}>
              {Array(count)
                .fill(0)
                .map((_, i) => urls[i % urls.length])
                .map((src, i) => (
                  <CarouselSlide key={i}>
                    <img
                      src={src}
                      alt={src}
                      style={{
                        width: '100%',
                        height: '100%',
                        objectFit: 'cover',
                        objectPosition: 'center center',
                      }}
                    />
                  </CarouselSlide>
                ))}
            </Carousel>
          </React.Fragment>
        ));
      })()}
    </div>
  );
};
