// @universe-design/react/src/carousel/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Basic from '../demo/0.basic';
import C1IndicatorType from '../demo/1.indicator-type';
import C2IndicatorTheme from '../demo/2.indicator-theme';
import C3_1SwitcherSize from '../demo/3.1.switcher-size';
import C3IndicatorPosition from '../demo/3.indicator-position';
import C4Controlled from '../demo/4.controlled';
import C5Duration from '../demo/5.duration';
import C6AutoPlay from '../demo/6.autoPlay';
import C7VariableWidth from '../demo/7.variable-width';
import C8Transition from '../demo/8.transition';
if (!(C0Basic as Story).storyName) {
  (C0Basic as Story).storyName = '0. 基础用法';
  (C0Basic as Story).parameters = {
    docs: {
      description: {
        story:
          '基本用法。由于轮播图组件幻灯片完全采用 `absolute` 布局，子元素高度并不能撑开父级，所以需要给父级元素单独设置高度，`CarouselSlide` 会默认占满整个轮播图窗口。\n\n',
      },
    },
  };
}
if (!(C1IndicatorType as Story).storyName) {
  (C1IndicatorType as Story).storyName = '1. 指示器类型';
  (C1IndicatorType as Story).parameters = {
    docs: {
      description: {
        story: '指示器有两种类型，分别是 `dot` 和 `line`，默认为 `dot`。\n\n',
      },
    },
  };
}
if (!(C2IndicatorTheme as Story).storyName) {
  (C2IndicatorTheme as Story).storyName = '2. 指示器主题';
  (C2IndicatorTheme as Story).parameters = {
    docs: {
      description: {
        story: '指示器主题有两种，分别是 `light` 和 `primary`，默认为 `light`。\n\n',
      },
    },
  };
}
if (!(C3_1SwitcherSize as Story).storyName) {
  (C3_1SwitcherSize as Story).storyName = '3.1. 翻页切换器尺寸';
  (C3_1SwitcherSize as Story).parameters = {
    docs: {
      description: {
        story: '翻页切换器共有两种尺寸 `small` 和 `medium`，默认是 `medium`。\n\n',
      },
    },
  };
}
if (!(C3IndicatorPosition as Story).storyName) {
  (C3IndicatorPosition as Story).storyName = '3. 指示器位置';
  (C3IndicatorPosition as Story).parameters = {
    docs: {
      description: {
        story:
          '指示器有两个位置，内部底部 `inner-bottom` 和外部底部 `outer-bottom`。默认为内部底部。指示器位置会影响指示器主题，外部指示器默认使用 `primary` 主题，内部指示器默认使用 `light` 主题。\n\n',
      },
    },
  };
}
if (!(C4Controlled as Story).storyName) {
  (C4Controlled as Story).storyName = '4. 受控模式';
  (C4Controlled as Story).parameters = {
    docs: {
      description: {
        story: '使用 `slideIndex` 和 `onSlideIndexChange` 实现受控模式。\n\n',
      },
    },
  };
}
if (!(C5Duration as Story).storyName) {
  (C5Duration as Story).storyName = '5. 动画持续时间';
  (C5Duration as Story).parameters = {
    docs: {
      description: {
        story: '转场动画的持续时间可以通过 `duration` 属性进行调整。\n\n',
      },
    },
  };
}
if (!(C6AutoPlay as Story).storyName) {
  (C6AutoPlay as Story).storyName = '6. 自动播放';
  (C6AutoPlay as Story).parameters = {
    docs: {
      description: {
        story: '自动播放速率可以通过 `autoPlay.interval` 来设置。\n\n',
      },
    },
  };
}
if (!(C7VariableWidth as Story).storyName) {
  (C7VariableWidth as Story).storyName = '7. 「幻灯片不定宽」轮播图';
  (C7VariableWidth as Story).parameters = {
    docs: {
      description: {
        story:
          '将 `fullWidth` 设置为 false，幻灯片宽度将由子元素决定，或直接给 `CarouselSlide` 设置宽度。此外，我们内置了幻灯片的 `ResizeObserver` 逻辑，动态加载导致元素尺寸变化也可以很好运行。\n\n',
      },
    },
  };
}
if (!(C8Transition as Story).storyName) {
  (C8Transition as Story).storyName = '8. 转场动画';
  (C8Transition as Story).parameters = {
    docs: {
      description: {
        story: '转场动画分为两种：「推移」`push` 和「淡入淡出」`fade`，默认为 `push`。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Carousel/官网示例',
  tags: ['autodocs'],
};

export {
  C0Basic,
  C1IndicatorType,
  C2IndicatorTheme,
  C3_1SwitcherSize,
  C3IndicatorPosition,
  C4Controlled,
  C5Duration,
  C6AutoPlay,
  C7VariableWidth,
  C8Transition,
};
