// @universe-design/react/src/util/assert.ts
import type { DetailedReactHTMLElement, InputHTMLAttributes, ReactElement, HTMLAttributes } from 'react';
import { isFunction, isElement } from '@universe-design/react-utilities';

export { isFunction } from '@universe-design/react-utilities';
export function isWindow(node: Node | typeof window): node is typeof window {
  return node.toString() === '[object Window]';
}

export function getWindow(node: Node | typeof window | null): typeof window {
  if (!node) {
    return window;
  }

  if (!isWindow(node)) {
    const { ownerDocument } = node;
    return ownerDocument ? ownerDocument.defaultView || window : window;
  }

  return node;
}

export function isNonNullable<T>(v: T): v is NonNullable<T> {
  if (v !== undefined && v !== null) {
    return true;
  }
  return false;
}

/** @internal is the given object an Object? */
export const isObject = (obj: unknown): obj is Record<string, unknown> => obj !== null && typeof obj === 'object';

/** @internal is the given object/value a promise? */
export const isPromise = (value: unknown): value is PromiseLike<unknown> => isObject(value) && isFunction(value.then);

export function isReactInputElement(
  node: ReactElement<any, any>,
): node is DetailedReactHTMLElement<InputHTMLAttributes<HTMLInputElement>, HTMLInputElement> {
  return node.type === 'input';
}

export function isReactHTMLElement(
  node: ReactElement<any, any>,
): node is DetailedReactHTMLElement<HTMLAttributes<HTMLElement>, HTMLElement> {
  return typeof node.type === 'string';
}

export const isString = (val: unknown): val is string => typeof val === 'string';

/**
 * Fork from https://github.com/juggle/resize-observer/blob/c92bcde23f23a5bc7b6d6e32449dec1e540c419f/src/ResizeObservation.ts#L6
 */
export const isSVG = (target: Element): boolean => target instanceof SVGElement && 'getBBox' in target;

export const isReplacedElement = (target: Element): boolean => {
  switch (target.tagName) {
    case 'INPUT':
      if ((target as HTMLInputElement).type !== 'image') {
        break;
      }
    case 'VIDEO':
    case 'AUDIO':
    case 'EMBED':
    case 'OBJECT':
    case 'CANVAS':
    case 'IFRAME':
    case 'IMG':
      return true;
    default:
  }
  return false;
};

export const isNonReplacedElement = (target: Element): boolean =>
  isElement(target) && !isSVG(target) && !isReplacedElement(target) && getComputedStyle(target).display === 'inline';
