// @universe-design/react/src/util/dom.ts
import { isElement, isHTMLElement } from '@universe-design/react-utilities';

export function getOwnerDocument(node?: Element | null): Document {
  return isElement(node) ? node.ownerDocument ?? document : document;
}

export function isInputElement(el: any): el is HTMLInputElement {
  return isHTMLElement(el) && el.tagName.toLowerCase() === 'input' && 'select' in el;
}

export function contains(parent: Node | null | undefined, child: Node) {
  if (!parent) {
    return false;
  }
  if (parent === child) {
    return true;
  }
  if ('contains' in parent) {
    return parent.contains(child);
  }
  return false;
}

export function eventContains(parent: Node | null | undefined, ev: Event) {
  if (!parent) {
    return false;
  }
  const { target } = ev;
  if (!isElement(target)) {
    return false;
  }
  if (ev.composedPath) {
    // 处理 shadowRoot 情况
    return ev.composedPath().indexOf(parent) !== -1;
  }
  return contains(parent, target);
}

export function getActiveElement(node?: HTMLElement) {
  const doc = getOwnerDocument(node);
  return doc?.activeElement as HTMLElement;
}

export function getRelatedTarget(event: Pick<FocusEvent, 'relatedTarget' | 'target' | 'currentTarget'>) {
  const target = (event.target ?? event.currentTarget) as HTMLElement;
  const activeElement = getActiveElement(target);
  return (event.relatedTarget ?? activeElement) as HTMLElement;
}

export function isHorizontalOverflowed(element: Element) {
  // 这个函数用于判断单行文本是否溢出，只需要判断 x 方向即可
  return element.scrollWidth > element.clientWidth;
}

export const isVerticalOverflowed = (ele: Element) =>
  // 这个函数用于判断多行文本是否溢出，只需要考虑 y 方向
  ele.clientHeight < ele.scrollHeight;

export function isOverflowed(element: Element) {
  // 默认的溢出判断逻辑，包含 x 和 y 两个方向，虽然在用户 line-height 设置有问题的时候可能会有一些误判，但是我们认为用户 line-height 小于内容高度是一种「错误」。
  return isHorizontalOverflowed(element) || isVerticalOverflowed(element);
}
