// @universe-design/react/src/util/focus.ts
// refer to https://github.com/chakra-ui/chakra-ui/blob/b577eac5d127b2e4cbedc2a15c2b0fcda03d7a17/packages/utils/src/focus.ts

import type { ThunkElement } from './getThunkElement';
import { getThunkElement } from './getThunkElement';
import { isInputElement } from './dom';
import type { FocusableElement } from './tabbable';
import { isActiveElement } from './tabbable';
import { devWarning, warning } from './warning';

let supportsPreventScrollCached: boolean | null = null;
function supportsPreventScroll() {
  if (supportsPreventScrollCached === null) {
    supportsPreventScrollCached = false;
    try {
      const div = document.createElement('div');
      div.focus({
        get preventScroll() {
          supportsPreventScrollCached = true;
          return true;
        },
      });
    } catch (e) {
      // Ignore
    }
  }

  return supportsPreventScrollCached;
}

export interface ExtendedFocusOptions extends FocusOptions {
  /**
   * If true, the element will be focused in the next tick
   */
  nextTick?: boolean;
  /**
   * If true and element is an input element, the input's text will be selected
   */
  selectTextIfInput?: boolean;
}

export function focus(thunkElement: ThunkElement<FocusableElement>, options: ExtendedFocusOptions = {}) {
  const isActive = isActiveElement;
  const { nextTick, preventScroll = true, selectTextIfInput = false } = options;

  function triggerFocus() {
    const element = getThunkElement(thunkElement);

    if (!element) {
      warning(false, devWarning('focus', "can't call focus() on `null` or `undefined` element"));
      return;
    }

    if (isActive(element)) {
      return;
    }

    if (supportsPreventScroll()) {
      element.focus({ preventScroll });
    } else {
      element.focus();
    }

    if (isInputElement(element) && selectTextIfInput) {
      element.select();
    }
  }

  if (nextTick) {
    return requestAnimationFrame(triggerFocus);
  }

  triggerFocus();
}
