// @universe-design/react/src/util/forwardRef.ts
/* eslint-disable no-undef */
import * as React from 'react';

export type IntrinsicElementTags = keyof JSX.IntrinsicElements;

export type InferHTMLElement<T extends IntrinsicElementTags> = JSX.IntrinsicElements[T] extends React.ClassAttributes<
  infer U
>
  ? U
  : never;

export const forwardHTMLRef = React.forwardRef as <
  P,
  T extends IntrinsicElementTags,
  RP = Omit<React.ComponentPropsWithoutRef<T>, keyof P> & P,
  RT = InferHTMLElement<T>,
>(
  render: React.ForwardRefRenderFunction<RT, RP>,
) => React.ForwardRefExoticComponent<React.PropsWithoutRef<RP> & React.RefAttributes<RT>>;

type ComponentPropsWithAS<C extends IntrinsicElementTags> = {
  as: C;
} & JSX.IntrinsicElements[C];

/**
 * @typeParam Props - 组件属性
 * @typeParam DefaultComp - 默认使用的组件
 * @typeParam Tags - 可用的组件
 * @typeParam Extra - 组件的额外属性
 */
export const genericForwardRef = React.forwardRef as <
  Props,
  DefaultComp extends Tags,
  Tags extends IntrinsicElementTags = IntrinsicElementTags,
  Extra = Record<string, never>,
>(
  render: (
    // 这里类型稍微宽松一点，实现的时候容易一些
    props: Partial<ComponentPropsWithAS<DefaultComp>> & Props,
    ref: React.ForwardedRef<InferHTMLElement<DefaultComp>>,
  ) => React.ReactElement | null,
) => {
  /**
   * 这里的类型设计思路如下：
   *
   * 1. 由于目前 ts (4.6) 仍然不能很好得支持 optional prop 的 type switch（{ type?: undefined } | { type: 'a' }），所以我们需要将缺省 `as` 的情况做成函数重载
   * 2. 由于 required prop 的 type switch 已经很稳定，所以有 `as` 的情况可以通过对 keyof IntrinsicElements 的 union type 做 mapping 得到带 as 的 html props
   * 3. 尝试过使用泛型 `<C extends IntrinsicElementTags>(props: ComponentPropsWithAS<C>): React.ReactElement | null` 去实现 2，但是发现这种实现会导致 event handler 和 ref 的 parameters 无法正常推断，所有最终还是采用了 props union type 的方式去实现了
   */
  (props: (Tags extends any ? ComponentPropsWithAS<Tags> : never) & Props): React.ReactElement | null;
  // eslint-disable-next-line @typescript-eslint/unified-signatures
  (props: JSX.IntrinsicElements[DefaultComp] & Props): React.ReactElement | null;
  displayName?: string;
} & Extra;
