// @universe-design/react/src/util/scrollTo.ts
/* eslint-disable no-redeclare */
import { isWindow } from './assert';
import { Cubic } from './easings';

type ContainerType = HTMLElement | typeof window | Document | undefined;
interface ScrollToOptions {
  getContainer?: () => ContainerType;
  scrollEnd?: () => void;
  duration?: number;
}

export function getSetScrollTop(target: ContainerType): number;
export function getSetScrollTop(target: ContainerType, scrollTop: number): void;
export function getSetScrollTop(target: ContainerType, scrollTop?: number) {
  if (!target) {
    return 0;
  }
  if (isWindow(target)) {
    if (scrollTop === undefined) {
      return target.pageYOffset;
    }
    target.scrollTo(window.pageXOffset, scrollTop);
    return void 0;
  }
  const el = target instanceof Document ? target.documentElement : target;
  if (scrollTop === undefined) {
    return el.scrollTop;
  }
  el.scrollTop = scrollTop;
  return void 0;
}

export function scrollTo(target: HTMLElement, mode: 'toTop' | 'toBottom') {
  if (target.parentElement) {
    const parentTop = target.parentElement.scrollTop;
    const parentBottom = parentTop + target.parentElement.clientHeight;
    const targetTop = target.offsetTop;
    const targetBottom = targetTop + target.clientHeight;
    if (targetTop < parentTop || targetBottom > parentBottom) {
      if (mode === 'toTop') {
        target.parentElement.scrollTop = target.offsetTop;
      } else if (mode === 'toBottom') {
        target.parentElement.scrollTop = targetBottom - target.parentElement.clientHeight;
      }
    }
  }
}

export function animationScrollTo(end: number, options: ScrollToOptions = {}) {
  const { getContainer = () => window, scrollEnd, duration = 450 } = options;

  const container = getContainer();
  const start = getSetScrollTop(container);

  const startTime = Date.now();
  let activated = true;
  let currentID: number;

  const doScroll = () => {
    if (activated === false) {
      return;
    }

    const timestamp = Date.now();
    const time = timestamp - startTime;
    const currentTime = time > duration ? duration : time;
    const nextTop = Cubic.inOut(currentTime, start, end, duration);
    getSetScrollTop(container, nextTop);

    if (time < duration) {
      currentID = requestAnimationFrame(doScroll);
    } else {
      scrollEnd && scrollEnd();
    }
  };

  currentID = requestAnimationFrame(doScroll);

  return () => {
    activated = false;
    cancelAnimationFrame(currentID);
  };
}
