// @universe-design/react/src/util/size.ts
import type { DefaultExtraSizeType, DefaultSizeType } from '@universe-design/react-utilities';
import type { ElementOf } from './types';

export const SizeEnum = ['xl', 'lg', 'md', 'sm', 'xs'] as const;

export type SizeType = ElementOf<typeof SizeEnum>;

/**
 * @public
 */
export const DefaultSizeEnum = ['large', 'medium', 'small'] as const;

export const DefaultExtraSizeEnum = [
  'extra-large',
  'extra-small',
  // 'default',
] as const;

export const sizeMap = {
  'extra-large': 'xl',
  large: 'lg',
  medium: 'md',
  // default: 'md',
  small: 'sm',
  'extra-small': 'xs',
} as const;

export function isValidDefaultSize(sizeName: string): sizeName is DefaultSizeType {
  return DefaultSizeEnum.includes(sizeName as DefaultSizeType);
}

export function isValidSize(sizeName: string): sizeName is DefaultSizeType | DefaultExtraSizeType {
  return (
    DefaultSizeEnum.includes(sizeName as DefaultSizeType) ||
    DefaultExtraSizeEnum.includes(sizeName as DefaultExtraSizeType)
  );
}

export function getDefaultSizeAbbr(sizeName: DefaultSizeType, defaultName?: DefaultSizeType) {
  if (isValidDefaultSize(sizeName)) {
    return sizeMap[sizeName];
  }
  if (defaultName && isValidDefaultSize(sizeName)) {
    return sizeMap[defaultName];
  }
  return '';
}

export type { DefaultExtraSizeType, DefaultSizeType } from '@universe-design/react-utilities';
export { getSizeAbbr } from '@universe-design/react-utilities';
