// @universe-design/react-aurora/lib-esm/aurora-background/AuroraBlock.js
import _extends from "@babel/runtime/helpers/esm/extends";
import produce from 'immer';
import React from 'react';
import { Ellipse } from './Ellipse';
import { Default_Animation_Time, Default_Aurora_Config } from './defaults';
import clsx from 'clsx';
import './styles/gradient-block.less';

/**
 * @public
 *  极光背景组件的属性
 */

/**
 * @public
 */
export const AuroraBlock = /*#__PURE__*/React.forwardRef((props, ref) => {
  const {
    mainEllipse: mainEllipseProp,
    subEllipse: subEllipseProp,
    reflectEllipse: reflectEllipseProp,
    speed = 1,
    noAnimation = false,
    className,
    ...rest
  } = props;
  const ellipseConfig = React.useMemo(() => ({
    ...Default_Aurora_Config,
    mainEllipse: {
      ...Default_Aurora_Config.mainEllipse,
      ...mainEllipseProp
    },
    subEllipse: {
      ...Default_Aurora_Config.subEllipse,
      ...subEllipseProp
    },
    reflectEllipse: {
      ...Default_Aurora_Config.reflectEllipse,
      ...reflectEllipseProp
    }
  }), [mainEllipseProp, reflectEllipseProp, subEllipseProp]);
  const [ellipse, setEllipse] = React.useState(ellipseConfig);
  const cancelId = React.useRef();
  const firstUpdateTimeRef = React.useRef();
  const updatePosition = (_ellipse, zoomRatio) => ({
    ..._ellipse,
    rX: _ellipse.rX + _ellipse.deltaRX * zoomRatio,
    rY: _ellipse.rY + _ellipse.deltaRY * zoomRatio
  });
  React.useLayoutEffect(() => {
    if (noAnimation) {
      return;
    }
    let zoomRatio = 0;
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    cancelId.current = window.setInterval(() => {
      const now = Date.now();
      if (firstUpdateTimeRef.current === undefined) {
        firstUpdateTimeRef.current = now;
      } else {
        const elapsed = now - firstUpdateTimeRef.current;
        zoomRatio = Math.sin(elapsed / Default_Animation_Time * 2 * Math.PI * speed);
      }
      setEllipse(produce(draft => {
        draft.mainEllipse = updatePosition(ellipseConfig.mainEllipse, zoomRatio);
        draft.subEllipse = updatePosition(ellipseConfig.subEllipse, zoomRatio);
        draft.reflectEllipse = updatePosition(ellipseConfig.reflectEllipse, zoomRatio);
      }));
    }, 16);
    return () => {
      cancelId.current && window.clearInterval(cancelId.current);
    };
  }, [speed, ellipseConfig.mainEllipse, ellipseConfig.subEllipse, ellipseConfig.reflectEllipse, noAnimation]);
  return /*#__PURE__*/React.createElement("svg", _extends({
    className: clsx(className, 'udx__aurora-block'),
    viewBox: "0 0 375 250",
    ref: ref
  }, rest), /*#__PURE__*/React.createElement(Ellipse, _extends({}, ellipse.mainEllipse, {
    filter: "url(#udx_aurora_main)"
  })), /*#__PURE__*/React.createElement(Ellipse, _extends({}, ellipse.subEllipse, {
    filter: "url(#udx_aurora_sub)"
  })), /*#__PURE__*/React.createElement(Ellipse, _extends({}, ellipse.reflectEllipse, {
    filter: "url(#udx_aurora_reflection)"
  })), /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("filter", {
    id: "udx_aurora_main",
    filterUnits: "userSpaceOnUse",
    "color-interpolation-filters": "sRGB"
  }, /*#__PURE__*/React.createElement("feFlood", {
    "flood-opacity": "0",
    result: "BackgroundImageFix"
  }), /*#__PURE__*/React.createElement("feBlend", {
    mode: "normal",
    in: "SourceGraphic",
    in2: "BackgroundImageFix",
    result: "shape"
  }), /*#__PURE__*/React.createElement("feGaussianBlur", {
    stdDeviation: "40",
    result: "effect1_foregroundBlur_210_68150"
  })), /*#__PURE__*/React.createElement("filter", {
    id: "udx_aurora_sub",
    filterUnits: "userSpaceOnUse",
    "color-interpolation-filters": "sRGB"
  }, /*#__PURE__*/React.createElement("feFlood", {
    "flood-opacity": "0",
    result: "BackgroundImageFix"
  }), /*#__PURE__*/React.createElement("feBlend", {
    mode: "normal",
    in: "SourceGraphic",
    in2: "BackgroundImageFix",
    result: "shape"
  }), /*#__PURE__*/React.createElement("feGaussianBlur", {
    stdDeviation: "50",
    result: "effect1_foregroundBlur_210_68150"
  })), /*#__PURE__*/React.createElement("filter", {
    id: "udx_aurora_reflection",
    filterUnits: "userSpaceOnUse",
    "color-interpolation-filters": "sRGB"
  }, /*#__PURE__*/React.createElement("feFlood", {
    "flood-opacity": "0",
    result: "BackgroundImageFix"
  }), /*#__PURE__*/React.createElement("feBlend", {
    mode: "normal",
    in: "SourceGraphic",
    in2: "BackgroundImageFix",
    result: "shape"
  }), /*#__PURE__*/React.createElement("feGaussianBlur", {
    stdDeviation: "40",
    result: "effect1_foregroundBlur_210_68150"
  }))));
});
