// @universe-design/react/src/menu/internal-menu/hooks/useActive.ts
import * as React from 'react';
import { RcMenuContext } from '../context/MenuContext';
import type { MenuHoverEventHandler } from '../interface';

interface ActiveObj {
  active: boolean;
  onMouseEnter?: React.MouseEventHandler<HTMLElement>;
  onMouseLeave?: React.MouseEventHandler<HTMLElement>;
}

export default function useActive(
  eventKey: string,
  disabled: boolean,
  onMouseEnter?: MenuHoverEventHandler,
  onMouseLeave?: MenuHoverEventHandler,
): ActiveObj {
  const {
    // Active
    activeKey,
    onActive,
    onInactive,
  } = React.useContext(RcMenuContext)!;

  const ret: ActiveObj = {
    // eventKey !== undefined 是为了修复初始化时，所有 eventKey 为空的 item 都会显示成 active 样式
    active: activeKey === eventKey && eventKey !== undefined,
  };

  // Skip when disabled
  if (!disabled) {
    ret.onMouseEnter = domEvent => {
      onMouseEnter?.({
        key: eventKey,
        domEvent,
      });
      onActive(eventKey);
    };
    ret.onMouseLeave = domEvent => {
      onMouseLeave?.({
        key: eventKey,
        domEvent,
      });
      onInactive(eventKey);
    };
  }

  return ret;
}
