// @byted-nodex/logger-core/lib/logger/core_logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable @typescript-eslint/no-explicit-any */
const util_1 = require("util");
const helper_1 = require("../util/helper");
const meta_1 = require("../meta");
const level_1 = require("../level");
const formatter_1 = require("../util/formatter");
class CoreLogger {
    constructor(options = {}) {
        this.version = meta_1.version;
        this.levels = level_1.LevelValues;
        this.flushTimer = null;
        this.name = options.name || this.constructor.name;
        this.level = options.level || level_1.LevelLabel.INFO;
        this.transports = new Map();
        this.formatData = options.formatData || formatter_1.textFormatter;
        this.transformData = options.transformData || (data => data);
        this.callSiteFrameIndex = options.callSiteFrameIndex || 3;
        this.flushInterval = options.flushInterval || 1500;
        if (options.transports && options.transports.length > 0) {
            this.addTransport(options.transports);
        }
    }
    addTransport(nameOrTransports, transport) {
        if (typeof nameOrTransports == 'string' && transport) {
            if (this.getTransport(nameOrTransports)) {
                throw new Error(`transport "${nameOrTransports}" has exists.`);
            }
            // 当往 transports map 对象添加一个新的 Transport 实例时，如果新增 transport 对象没有设置 level 信息，则默认设置成 Logger 实例 level
            if (!transport.getLevel()) {
                transport.setLevel(this.level);
            }
            this.transports.set(nameOrTransports, transport);
        }
        else if (Array.isArray(nameOrTransports)) {
            nameOrTransports.forEach(transport => {
                this.addTransport(transport.name, transport);
            });
        }
        else {
            throw new Error('param error.');
        }
        return this;
    }
    removeTransport(name) {
        const transport = this.getTransport(name);
        if (transport) {
            transport.close();
            this.transports.delete(name);
        }
    }
    getTransport(name) {
        return this.transports.get(name);
    }
    disableTransport(name) {
        var _a;
        (_a = this.transports.get(name)) === null || _a === void 0 ? void 0 : _a.disable();
    }
    enableTransport(name) {
        var _a;
        (_a = this.transports.get(name)) === null || _a === void 0 ? void 0 : _a.enable();
    }
    reloadTransport(name) {
        var _a;
        (_a = this.getTransport(name)) === null || _a === void 0 ? void 0 : _a.reload();
    }
    reloadAllTransports() {
        for (const transport of this.transports.values()) {
            transport.reload();
        }
    }
    closeTransport(name) {
        var _a;
        (_a = this.getTransport(name)) === null || _a === void 0 ? void 0 : _a.close();
    }
    closeAllTransports() {
        for (const transport of this.transports.values()) {
            transport.close();
        }
    }
    /**
     * Send log to all transports
     */
    log(data) {
        if (level_1.LevelValues[data.level] < level_1.LevelValues[this.level]) {
            return;
        }
        if (!this.flushTimer) {
            this.flushTimer = setTimeout(() => {
                this.flush();
            }, this.flushInterval).unref();
        }
        // default data transform
        const { timestamp, datetimeFormatted } = (0, helper_1.datetime)();
        data.pid = meta_1.pid;
        data.version = meta_1.version;
        data.timestamp = timestamp;
        data.datetime = datetimeFormatted;
        data.hostIP = meta_1.hostIP;
        data.ipv4 = meta_1.hostIPV4;
        data.ipv6 = meta_1.hostIPV6;
        data.tags = data.tags || {};
        data.location = data.location || (0, helper_1.getStackFilePath)(this.callSiteFrameIndex);
        data.msg = data.msg || '';
        if (data.msg instanceof Error) {
            data.msg = data.msg.stack;
        }
        else if (data.msg && data.splats) {
            data.msg = (0, util_1.format)(data.msg, ...data.splats);
        }
        const transformedData = this.transformData(data);
        const formattedData = this.formatData(transformedData);
        for (const transport of this.transports.values()) {
            if (transport.shouldLog(data.level)) {
                transport.log(formattedData, transformedData);
            }
        }
    }
    flush() {
        for (const transport of this.transports.values()) {
            transport.flush();
        }
        this.flushTimer = null;
    }
    flushSync() {
        for (const transport of this.transports.values()) {
            transport.flushSync();
        }
    }
    close() {
        this.flushSync();
        if (this.flushTimer) {
            clearTimeout(this.flushTimer);
            this.flushTimer = null;
        }
        this.closeAllTransports();
    }
    trace(msg, ...splats) {
        return this.log({ level: level_1.LevelLabel.TRACE, msg, splats });
    }
    debug(msg, ...splats) {
        return this.log({ level: level_1.LevelLabel.DEBUG, msg, splats });
    }
    info(msg, ...splats) {
        return this.log({ level: level_1.LevelLabel.INFO, msg, splats });
    }
    warn(msg, ...splats) {
        return this.log({ level: level_1.LevelLabel.WARN, msg, splats });
    }
    error(msg, ...splats) {
        return this.log({ level: level_1.LevelLabel.ERROR, msg, splats });
    }
    fatal(msg, ...splats) {
        return this.log({ level: level_1.LevelLabel.FATAL, msg, splats });
    }
}
exports.default = CoreLogger;
CoreLogger.version = meta_1.version;
