// @universe-design/react/lib/date-picker/panels/month-body/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _Body = _interopRequireDefault(require("../base/Body"));
var _config = require("../base/config");
var _hooks = require("../../../config-provider/hooks");
var _cell = require("../../cell");
var _DateCalendarUIContext = require("../DateCalendar/DateCalendarUIContext");
var _utils = require("../DateCalendar/utils");
function MonthPanel() {
  const locale = (0, _hooks.useLocaleReceiver)('datePicker');
  const {
    shouldCellDisabled,
    cellRenderer,
    pickerValue,
    selectedDates,
    selectedRange,
    previewDate
  } = (0, _DateCalendarUIContext.useDateCalendarUIContext)();
  const prefixCls = (0, _hooks.useCssScope)('picker')();
  const dateCellInteractiveAreaCls = (0, _cell.useDateCellInteractiveAreaCls)();
  return /*#__PURE__*/_react.default.createElement(_Body.default, (0, _extends2.default)({}, _config.BodyConfig.month, {
    cellClassName: `${`${prefixCls}-month-panel`}-cell`,
    baseDate: pickerValue.clone().startOf('year'),
    getCellStatus: date => {
      const isRangePreview = previewDate ? previewDate(date) : false;
      const [startDate = null, endDate = null] = selectedRange || [];
      return {
        isDisabled: shouldCellDisabled(date, 'month'),
        isRange: date.isSameOrAfter(startDate, 'month') && date.isSameOrBefore(endDate, 'month'),
        isSelected: (0, _utils.isSelectCell)({
          selectedDates,
          selectedRange,
          mode: 'month',
          date
        }),
        isRangePreview
      };
    },
    getCellDate: (d, offset) => d.clone().add(offset, 'month'),
    dateRender: d => {
      const node = /*#__PURE__*/_react.default.createElement("div", {
        className: dateCellInteractiveAreaCls
      }, locale.monthsMin[d.month()]);
      if (typeof cellRenderer === 'function') {
        return cellRenderer(d, {
          viewMode: 'month',
          defaultNode: node
        });
      }
      return node;
    }
  }));
}
var _default = exports.default = MonthPanel;
