// @universe-design/react/lib/layout/Layout.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.LayoutContext = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _hooks = require("../config-provider/hooks");
require("./style");
const LayoutContext = exports.LayoutContext = /*#__PURE__*/_react.default.createContext({
  siderControl: {
    addSider: () => null,
    removeSider: () => null
  }
});
const Layout = /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
  const {
    className,
    hasSider,
    children,
    ...restProps
  } = props;
  const [siders, setSiders] = (0, _react.useState)([]);
  const cssScope = (0, _hooks.useCssScope)('layout');
  const cls = (0, _clsx.default)({
    [cssScope()]: true,
    [cssScope('--has-sider')]: !(hasSider === false) && (hasSider || siders.length > 0)
  }, className);
  return /*#__PURE__*/_react.default.createElement(LayoutContext.Provider, {
    value: {
      siderControl: (() => ({
        addSider: id => {
          setSiders(prev => [...prev, id]);
        },
        removeSider: id => {
          setSiders(prev => prev.filter(currentId => currentId !== id));
        }
      }))()
    }
  }, /*#__PURE__*/_react.default.createElement("section", (0, _extends2.default)({
    ref: ref,
    className: cls
  }, restProps), children));
});
var _default = exports.default = Layout;
