// @universe-design/react/lib/layout/Sider.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.SiderContext = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _reactUtilities = require("@universe-design/react-utilities");
var _hooks = require("../hooks");
var _Layout = require("./Layout");
var _hooks2 = require("../config-provider/hooks");
var _util = require("../util");
var _SideExpandOutlined = _interopRequireDefault(require("@universe-design/icons-react/SideExpandOutlined"));
var _SideFoldOutlined = _interopRequireDefault(require("@universe-design/icons-react/SideFoldOutlined"));
var _DisordeListOutlined = _interopRequireDefault(require("@universe-design/icons-react/DisordeListOutlined"));
require("./style");
const dimensionMaxMap = {
  xs: '479.98px',
  sm: '575.98px',
  md: '767.98px',
  lg: '991.98px',
  xl: '1199.98px',
  xxl: '1599.98px'
};
const SiderContext = exports.SiderContext = /*#__PURE__*/_react.default.createContext(null);
if (process.env.NODE_ENV !== "production") {
  SiderContext.displayName = 'SiderContext';
}

/**
 * @public
 */

/**
 * @public
 * Layout.Sider 的属性和方法
 */

const generateId = (() => {
  let i = 0;
  return function () {
    let prefix = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
    i += 1;
    return `${prefix}${i}`;
  };
})();

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const isNumeric = value => !isNaN(parseFloat(value)) && isFinite(value);

// eslint-disable-next-line max-lines-per-function
const Sider = /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
  const {
    theme,
    style,
    className,
    width = 200,
    trigger,
    collapsed,
    collapsedWidth = 64,
    defaultCollapsed,
    collapsible,
    reverseArrow,
    zeroWidthTriggerStyle,
    breakpoint,
    children,
    onCollapse,
    onBreakpoint,
    ...othereProps
  } = props;
  const cssScope = (0, _hooks2.useCssScope)('layout__sider');
  const prefixCls = cssScope();
  const mqlRef = (0, _react.useRef)();
  const {
    siderControl
  } = (0, _react.useContext)(_Layout.LayoutContext);
  const [below, setBelow] = (0, _react.useState)(false);
  const [ignoreState, setIgnoreState] = (0, _react.useState)(false);
  const [collapsedState, setCollapsedState] = (0, _reactUtilities.useControlledState)(Boolean(defaultCollapsed), collapsed);
  const setCollapsed = (0, _react.useCallback)((collapsedValue, type) => {
    setCollapsedState(collapsedValue);
    (0, _util.isFunction)(onCollapse) && onCollapse(collapsedValue, type);
  }, [onCollapse, setCollapsedState]);
  const responsiveHandler = (0, _react.useCallback)(mql => {
    setBelow(mql.matches);
    (0, _util.isFunction)(onBreakpoint) && onBreakpoint(mql.matches);
    if (!mql.matches) {
      setIgnoreState(false);
    }
    if (!ignoreState && collapsedState !== mql.matches) {
      setCollapsed(mql.matches, 'responsive');
    }
  }, [collapsedState, ignoreState, onBreakpoint, setCollapsed, setIgnoreState]);
  const toggle = () => {
    setCollapsed(!collapsedState, 'clickTrigger');
    setIgnoreState(true);
  };
  if (!mqlRef.current && typeof window !== 'undefined' && window.matchMedia && breakpoint) {
    mqlRef.current = window.matchMedia(`(max-width: ${dimensionMaxMap[breakpoint]})`);
    (mqlRef.current.addEventListener || mqlRef.current.addListener)('change', responsiveHandler);
    responsiveHandler(mqlRef.current);
  }
  (0, _hooks.useEffectOnce)(() => {
    const uuid = generateId(cssScope());
    siderControl.addSider(uuid);
    return () => {
      siderControl.removeSider(uuid);
    };
  });
  (0, _react.useEffect)(() => {
    let mql = null;
    if (!mql && typeof window !== 'undefined' && window.matchMedia && breakpoint) {
      mql = window.matchMedia(`(max-width: ${dimensionMaxMap[breakpoint]})`);
      if (mql.addEventListener) {
        mql.addEventListener('change', responsiveHandler);
      } else {
        mql.addListener(responsiveHandler);
      }
      responsiveHandler(mql);
    }
    return () => {
      if (mql) {
        if (mql.removeEventListener) {
          mql.removeEventListener('change', responsiveHandler);
        } else {
          mql.removeListener(responsiveHandler);
        }
        mql = null;
      }
    };
  }, [breakpoint, responsiveHandler]);
  const rawWidth = collapsedState ? collapsedWidth : width;
  const siderWidth = isNumeric(rawWidth) ? `${rawWidth}px` : String(rawWidth);
  const zeroWidthTrigger = parseFloat(String(collapsedWidth || 0)) === 0 ? /*#__PURE__*/_react.default.createElement("span", {
    onClick: toggle,
    className: (0, _clsx.default)(cssScope('--zero-width-trigger'), cssScope(`--zero-width-trigger-${reverseArrow ? 'right' : 'left'}`)),
    style: zeroWidthTriggerStyle
  }, trigger || /*#__PURE__*/_react.default.createElement(_DisordeListOutlined.default, null)) : null;
  const SiderStyle = {
    ...style,
    flex: `0 0 ${siderWidth}`,
    maxWidth: siderWidth,
    minWidth: siderWidth,
    width: siderWidth,
    '--collapsedWidth': typeof collapsedWidth === 'number' ? `${collapsedWidth}px` : collapsedWidth
  };
  const siderCls = (0, _clsx.default)(className, prefixCls, {
    [cssScope(`__${theme}`)]: theme,
    [cssScope('--collapsed')]: Boolean(collapsedState),
    [cssScope('--has-trigger')]: collapsible && trigger !== null && !zeroWidthTrigger,
    [cssScope('--below')]: Boolean(below),
    [cssScope('--zero-width')]: parseFloat(siderWidth) === 0
  });
  const iconObj = {
    expanded: reverseArrow ? /*#__PURE__*/_react.default.createElement(_SideFoldOutlined.default, null) : /*#__PURE__*/_react.default.createElement(_SideExpandOutlined.default, null),
    collapsed: reverseArrow ? /*#__PURE__*/_react.default.createElement(_SideExpandOutlined.default, null) : /*#__PURE__*/_react.default.createElement(_SideFoldOutlined.default, null)
  };
  const status = collapsedState ? 'collapsed' : 'expanded';
  const defaultTrigger = iconObj[status];
  return /*#__PURE__*/_react.default.createElement(SiderContext.Provider, {
    value: {
      siderCollapsed: collapsedState,
      collapsedWidth
    }
  }, /*#__PURE__*/_react.default.createElement("aside", (0, _extends2.default)({
    ref: ref,
    style: SiderStyle,
    className: siderCls
  }, othereProps), /*#__PURE__*/_react.default.createElement("div", {
    className: cssScope('__children')
  }, children), (() => {
    if (collapsible || below && zeroWidthTrigger && trigger !== null) {
      return zeroWidthTrigger || /*#__PURE__*/_react.default.createElement("div", {
        className: `${prefixCls}-trigger`,
        onClick: toggle,
        style: {
          width: siderWidth
        }
      }, trigger || defaultTrigger);
    }
    return null;
  })()));
});
var _default = exports.default = Sider;
