// @universe-design/react/lib/cascader/cascade/CascadeContext.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useCascadeContext = exports.CascadeContextProvider = void 0;
var _reactUtilities = require("@universe-design/react-utilities");
// export type CascadeContextType = {
//   currentExpandKey: string | null;
//   setCurrentExpandKey: React.Dispatch<React.SetStateAction<string | null>>;
//   checkable: boolean;
//   multiple: boolean;
//   inputValue: string;
//   selected: string[];
//   labelRenderer: (node: DoublyTreeNode) => React.ReactNode;
//   setLoadedKeys: UseControlledStateSetState<string[]>;
//   searchFlattenTree: DoublyTreeNode[];
// } & ReturnType<typeof genDoublyTreeNodeMap> &
// Required<
// Omit<CascadeProps, 'options'>,
// 'listHeight' | 'changeOnSelect' | 'multiple' | 'dropdownLevelWidth'
// >;

const {
  Provider: CascadeContextProvider,
  useContext: useCascadeContext
} = (0, _reactUtilities.createContextHook)('CascaderContext', {
  strict: true
});
exports.useCascadeContext = useCascadeContext;
exports.CascadeContextProvider = CascadeContextProvider;
