// @universe-design/react/lib/cascader/cascade/Option.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.CascaderOption = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _ListCheckBoldOutlined = _interopRequireDefault(require("@universe-design/icons-react/ListCheckBoldOutlined"));
var _RightBoldOutlined = _interopRequireDefault(require("@universe-design/icons-react/RightBoldOutlined"));
var _clsx = _interopRequireDefault(require("clsx"));
var _stylex = require("../../stylex");
var _checkbox = require("../../checkbox");
var _tree = require("../../tree");
var _common = require("../common");
var _util = require("../../util");
var _composeRefs = _interopRequireDefault(require("../../util/composeRefs"));
var _textOverflow = require("../../shared/components/text-overflow");
var _useTreeNodeLoadData = require("../../tree/hooks/useTreeNodeLoadData");
var _loading = require("../../loading");
var _CascadeContext = require("./CascadeContext");
/* eslint-disable max-lines-per-function */
/* eslint-disable @typescript-eslint/no-non-null-assertion */

const CascaderOption = exports.CascaderOption = /*#__PURE__*/React.forwardRef((_ref, outerRef) => {
  let {
    isChecked,
    isSelected = false,
    node,
    showExpandIcon,
    highlightOption
  } = _ref;
  const {
    labelRenderer,
    loadedKeys,
    setLoadedKeys,
    onLoad,
    loadData,
    setCurrentExpandKey,
    expandTrigger,
    checkable,
    handleSelectChange,
    handleCheckChange
  } = (0, _CascadeContext.useCascadeContext)();
  const checkboxContainerRef = React.useRef(null);
  const {
    __node__,
    key
  } = node;
  const isHighlight = node === highlightOption && (0, _tree.isNonLeaf)(node.__node__);
  const {
    value,
    disabled = false
  } = __node__;
  const disableCheckbox = (0, _tree.isCheckedDisabled)(__node__);
  const cssScope = (0, _common.useCascaderMenuCssScope)();
  const {
    loading,
    onInternalExpandedToggle,
    isLeaf: isLoadableLeaf
  } = (0, _useTreeNodeLoadData.useTreeNodeLoadData)({
    doublyTreeNode: node,
    loadedKeys: loadedKeys || [],
    setLoadedKeys,
    onLoad,
    loadData,
    toggleExpanded: () => setCurrentExpandKey(node.key)
  });
  const waitToLoad = loadData && !(loadedKeys !== null && loadedKeys !== void 0 && loadedKeys.includes(key)) && !isLoadableLeaf;
  /**
   * ======= 自动 scroll 到该元素 ========
   */
  const ref = React.useRef(null);
  React.useEffect(() => {
    const target = ref.current;
    if (isHighlight && target && target.parentElement) {
      const parentTop = target.parentElement.scrollTop;
      const parentBottom = parentTop + target.parentElement.clientHeight;
      const targetTop = target.offsetTop;
      const targetBottom = targetTop + target.clientHeight;
      if (targetTop < parentTop || targetBottom > parentBottom) {
        target.parentElement.scrollTop = target.offsetTop;
      }
    }
  }, [isHighlight]);
  const handleMouseEnter = React.useCallback(() => {
    if (!node || expandTrigger !== 'hover') {
      return;
    }
    if (disabled) {
      return;
    }
    onInternalExpandedToggle();
  }, [disabled, expandTrigger, node, onInternalExpandedToggle]);
  const handleClick = React.useCallback(e => {
    const {
      target
    } = e;
    if ((target instanceof HTMLElement || target instanceof SVGElement) && checkboxContainerRef.current && checkboxContainerRef.current.contains(target)) {
      return;
    }
    if (disabled) {
      return;
    }
    if (waitToLoad) {
      onInternalExpandedToggle();
    } else {
      handleSelectChange(node);
    }
  }, [disabled, waitToLoad, onInternalExpandedToggle, handleSelectChange, node]);
  const handleCheckboxClick = React.useCallback(() => {
    handleCheckChange(value);
  }, [handleCheckChange, value]);
  const rightIconNode = React.useMemo(() => {
    // level 显示的时候非叶子节点显示 expand icon，其他选中态都是显示默认的 checker icon
    if (loading) {
      return /*#__PURE__*/React.createElement(_loading.Loading, {
        size: "small"
      });
    }
    if (!isLoadableLeaf && showExpandIcon) {
      return /*#__PURE__*/React.createElement(_RightBoldOutlined.default, {
        className: cssScope('__option__expandIcon')
      });
    }
    if (!checkable) {
      return /*#__PURE__*/React.createElement("span", {
        className: cssScope('__option__checker')
      }, isSelected && /*#__PURE__*/React.createElement(_ListCheckBoldOutlined.default, null));
    }
  }, [checkable, cssScope, isLoadableLeaf, isSelected, loading, showExpandIcon]);

  // ================ Checkbox ===============

  let checkboxProps = {
    checked: false,
    indeterminate: false,
    disabled: disableCheckbox
  };
  if (checkable) {
    if (isChecked === _tree.CheckedType.Checked) {
      checkboxProps = {
        checked: true,
        indeterminate: false,
        disabled: disableCheckbox
      };
    } else if (isChecked === _tree.CheckedType.Indeterminate) {
      checkboxProps = {
        checked: false,
        indeterminate: true,
        disabled: disableCheckbox
      };
    }
    if (isChecked === undefined) {
      _util.logger.error(`${__node__.value}, isChecked(${isChecked}) is undefined`);
      return null;
    }
  }
  return /*#__PURE__*/React.createElement("div", {
    ref: (0, _composeRefs.default)(ref, outerRef),
    onClick: handleClick,
    onMouseEnter: handleMouseEnter // 由于打开下拉菜单时有动画，可能会hover到第一个选项，以至于展开菜单
    ,
    key: value,
    className: (0, _clsx.default)(cssScope('__option'), (0, _tree.isLeaf)(__node__) && cssScope('__option-leaf'), isSelected && cssScope('__option-selected'), isHighlight && cssScope('__option-active'), disabled && cssScope('__option-disabled'))
  }, checkable && /*#__PURE__*/React.createElement(_checkbox.Checkbox, (0, _extends2.default)({
    ref: checkboxContainerRef,
    className: cssScope('__option__checkbox')
  }, checkboxProps, {
    onChange: handleCheckboxClick
  })), /*#__PURE__*/React.createElement(_textOverflow.TextOverflow, {
    as: "div",
    className: (0, _clsx.default)(cssScope('__option__label')),
    lineClamp: 1,
    sx: [_stylex.cssTypographyBody0]
  }, labelRenderer(node)), rightIconNode);
});
var _default = exports.default = CascaderOption;
