// @universe-design/react/lib/portal/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Portal = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var React = _interopRequireWildcard(require("react"));
var _reactDom = _interopRequireDefault(require("react-dom"));
var _clsx = _interopRequireDefault(require("clsx"));
var _util = require("../util");
var _portalContext = require("./portal-context");
var _hooks = require("../config-provider/hooks");
var _getThunkElement = require("../util/getThunkElement");
const {
  Provider: PortalContextProvider
} = _portalContext.PortalContext;
class PortalBase extends React.Component {
  constructor() {
    super(...arguments);
    (0, _defineProperty2.default)(this, "currentHost", null);
    (0, _defineProperty2.default)(this, "templateNode", null);
    (0, _defineProperty2.default)(this, "currentDocument", null);
    (0, _defineProperty2.default)(this, "portal", null);
    (0, _defineProperty2.default)(this, "currentWrapperCls", '');
    (0, _defineProperty2.default)(this, "firstRender", true);
    (0, _defineProperty2.default)(this, "getHost", void 0);
    (0, _defineProperty2.default)(this, "updateRef", () => {
      const {
        innerRef
      } = this.props;
      innerRef && ((0, _util.isFunction)(innerRef) && innerRef(this.portal) || typeof innerRef === 'object' && (innerRef.current = this.portal));
    });
    (0, _defineProperty2.default)(this, "updateCls", () => {
      const {
        innerWrapperCls,
        wrapperClassName
      } = this.props;
      if (this.currentWrapperCls !== wrapperClassName) {
        this.currentWrapperCls = wrapperClassName || '';
        this.portal && (this.portal.className = (0, _clsx.default)(innerWrapperCls, wrapperClassName));
      }
    });
    (0, _defineProperty2.default)(this, "innerGetHost", () => this.getHost(this.currentHost, this.templateNode));
    (0, _defineProperty2.default)(this, "upsertHost", () => {
      const nextHost = this.innerGetHost();
      if (nextHost !== this.currentHost) {
        const hostOnScreen = this.currentHost;
        this.currentHost = nextHost;
        if (hostOnScreen && this.portal) {
          // 从当前页面上存在的 host 里卸载之前的 portal
          hostOnScreen.contains(this.portal) && hostOnScreen.removeChild(this.portal);
        }
        if (!this.currentHost) {
          return;
        }
        const nextDocument = this.currentHost.ownerDocument;
        if (nextDocument && this.currentDocument !== nextDocument) {
          this.portal = nextDocument.createElement('div');
          this.currentDocument = nextDocument;
          // 初次渲染的时候可以不用执行 setState
          !this.firstRender && this.setState({});
          this.updateRef();
        }
        if (this.portal) {
          this.currentHost.appendChild(this.portal);
        }
      }
      this.firstRender && (this.firstRender = false);
    });
  }
  componentDidMount() {
    this.upsertHost();
    this.updateCls();
  }
  componentDidUpdate() {
    this.upsertHost();
    this.updateCls();
  }
  componentWillUnmount() {
    if (this.currentHost && this.portal) {
      this.currentHost.contains(this.portal) && this.currentHost.removeChild(this.portal);
    }
    this.portal = null;
    this.currentHost = null;
    this.currentDocument = null;
    this.templateNode = null;
  }
  render() {
    const {
      children
    } = this.props;
    this.upsertHost();
    return this.portal ? (/*#__PURE__*/_reactDom.default.createPortal(/*#__PURE__*/React.createElement(PortalContextProvider, {
      value: this.portal
    }, children), this.portal
    // span 的 ref 中只记录, 不清除 templateNode,
    // 是为了如果把 children 渲染出来了此时 span 没了
    // 但是会触发 componentDidUpdate
    // componentDidUpdate 里触发 getHost, 此时如果 templateNode 随着 span 消失而变成 null 的话
    // 那前后两次的 host 可能就不一样了
    )) : /*#__PURE__*/React.createElement("span", {
      ref: n => n && (this.templateNode = n)
    });
  }
}
class ContainerPortal extends PortalBase {
  constructor() {
    super(...arguments);
    (0, _defineProperty2.default)(this, "getHost", () => {
      const {
        container
      } = this.props;
      return (0, _getThunkElement.getThunkElement)(container);
    });
  }
}
class DefaultPortal extends PortalBase {
  constructor() {
    super(...arguments);
    (0, _defineProperty2.default)(this, "getHost", (currentHost, templateNode) => {
      const {
        appendToParentPortal = true,
        parentPortal = null,
        iframeRender = false
      } = this.props;
      if (!currentHost && (iframeRender || appendToParentPortal && parentPortal) && !templateNode) {
        /**
         * 几种情况需要让 host 是 null
         * 1. 当 iframeRender 为 false 时, 表示不用考虑渲染在 iframe 的场景
         *    因此这种情况下可以在第一次渲染的时候就直接拿到精确的 host
         *    但是在嵌套的场景中, 如果允许把 portal 渲染在父级 portal 上的话
         *    由于在渲染父级 portal 时已经把 context 中注入了 parentPortal
         *    所以这里若是不直接让 host = null 的话
         *    会导致之后直接走到 appendChild 的逻辑
         *    此时万一这个子 portal 有个哥哥节点的话
         *    那这个 portal 就会比它哥哥先被挂到父 portal 下, 顺序就乱了
         *
         * 2. 用户希望可以在 iframe 中渲染, 并且此时还没有临时的 span, 也就意味着此时无论如何也拿不到精准的 document
         */
        return null;
      }
      if (appendToParentPortal && parentPortal) {
        return parentPortal;
      }
      if (iframeRender && templateNode) {
        return templateNode.ownerDocument.body;
      }
      return document.body;
    });
  }
}
const Portal = exports.Portal = /*#__PURE__*/React.forwardRef((props, ref) => {
  const parentPortal = (0, _portalContext.usePortalContext)();
  const wrapperClsScope = (0, _hooks.useCssScope)('portal')();
  if (props.container === false) {
    return /*#__PURE__*/React.createElement(React.Fragment, null, props.children);
  }
  return props.container !== undefined && props.container !== null ? /*#__PURE__*/React.createElement(ContainerPortal, (0, _extends2.default)({}, props, {
    container: props.container,
    innerRef: ref,
    innerWrapperCls: wrapperClsScope
  })) : /*#__PURE__*/React.createElement(DefaultPortal, (0, _extends2.default)({}, props, {
    parentPortal: parentPortal,
    innerWrapperCls: wrapperClsScope,
    innerRef: ref
  }));
});
if (process.env.NODE_ENV !== "production") {
  Portal.displayName = 'Portal';
}
var _default = exports.default = Portal;
