// @universe-design/react/src/shared/components/transition/SlideTransition.tsx
/* eslint-disable @typescript-eslint/no-unused-vars */
import type React from 'react';
import type { CSSTransitionProps } from 'react-transition-group/CSSTransition';
import type { TransitionProps } from 'react-transition-group/Transition';
import { createInOutTransition } from '@universe-design/react-utilities';
import { transitionTimingFunctionStandard } from './timingFunctionConsts';
import { useCssScope } from '../../../config-provider/hooks';

export type SlideDirection = 'up' | 'right' | 'down' | 'left';

export const SlideTransition = createInOutTransition<{
  direction: SlideDirection;
  size: number | string;
}>({
  name: config => `slide-${config.direction}-${config.size}`,
  styles: config => {
    const translateName = ['up', 'down'].includes(config.direction) ? 'translateY' : 'translateX';
    const inStyle = `opacity: 0; transform: ${translateName}(${
      ['right', 'down'].includes(config.direction) ? '-' : ''
    }${typeof config.size === 'string' ? config.size : `${config.size}px`});`;
    const outStyle = `opacity:1; transform: ${translateName}(0px);`;
    return {
      appear: inStyle,
      enter: inStyle,
      exitDone: inStyle,
      appearActive: outStyle,
      enterActive: outStyle,
      exit: outStyle,
      exitActive: inStyle,
    };
  },
  transitions: {
    appearActive: t =>
      `opacity ${t}ms ${transitionTimingFunctionStandard}, transform ${t}ms ${transitionTimingFunctionStandard}`,
    enterActive: t =>
      `opacity ${t}ms ${transitionTimingFunctionStandard}, transform ${t}ms ${transitionTimingFunctionStandard}`,
    exitActive: t =>
      `opacity ${t}ms ${transitionTimingFunctionStandard}, transform ${t}ms ${transitionTimingFunctionStandard}`,
  },
  useCssScope,
});
