// @universe-design/react/src/shared/components/transition/SlideUpDownTransition.tsx
/* eslint-disable @typescript-eslint/no-unused-vars */
import type React from 'react';
import type { CSSTransitionProps } from 'react-transition-group/CSSTransition';
import type { TransitionProps } from 'react-transition-group/Transition';
import type { Placement } from '../popper/types';
import { createInOutTransition } from '@universe-design/react-utilities';
import { transitionTimingFunctionStandard } from './timingFunctionConsts';
import { useCssScope } from '../../../config-provider/hooks';

/**
 * 用于 select dropdown 等下拉
 */
export const SlideUpDownTransition = createInOutTransition<{
  placement: Placement;
}>({
  name: ({ placement }) => `slide-up-down-${placement}`,
  styles: ({ placement }) => {
    const [mainAxis, crossAxis] = placement.split('-');
    let translateY = '0';
    if (mainAxis === 'bottom' || (mainAxis.match(/right|left/) && crossAxis === 'start')) {
      translateY = '-8px';
    } else if (mainAxis === 'top' || (mainAxis.match(/right|left/) && crossAxis === 'end')) {
      translateY = '8px';
    }
    return {
      appear: `opacity: 0;transform: translateY(${translateY});`,
      appearActive: 'opacity: 1;transform: translateY(0);',
      appearDone: 'opacity: 1;transform: translateY(0);',
      enter: `opacity: 0;transform: translateY(${translateY});`,
      enterActive: 'opacity: 1;transform: translateY(0);',
      enterDone: 'opacity: 1;transform: translateY(0);',
      exit: 'opacity: 1;transform: translateY(0);pointer-events: none;',
      exitActive: `opacity: 0;transform: translateY(${translateY});`,
      exitDone: `opacity: 0;transform: translateY(${translateY});pointer-events: auto;`,
    };
  },
  transitions: {
    appearActive: t =>
      `opacity ${t}ms ${transitionTimingFunctionStandard}, transform ${t}ms ${transitionTimingFunctionStandard}`,
    enterActive: t =>
      `opacity ${t}ms ${transitionTimingFunctionStandard}, transform ${t}ms ${transitionTimingFunctionStandard}`,
    exitActive: t =>
      `opacity ${t}ms ${transitionTimingFunctionStandard}, transform ${t}ms ${transitionTimingFunctionStandard}`,
  },
  useCssScope,
});
