// @universe-design/react/src/shared/components/transition/useAnimatedVisible.ts
import React from 'react';
import type { CSSTransitionProps } from 'react-transition-group/CSSTransition';
import { isFunction } from '../../../util';
import { isClient } from '@universe-design/react-utilities';
import type { Required } from 'utility-types';

export interface UseAnimatedVisibleProps {
  /**
   * 是否在首次加载未显示的时候强制渲染
   * @defaultValue false
   */
  forceRender?: boolean;
  /**
   * 是否在关闭的时候卸载节点
   * @defaultValue false
   */
  destroyOnClose?: boolean;
  /**
   * 是否可见
   */
  visible: boolean;
  /**
   * 进入动画完成后的回调
   */
  onAfterOpen?: () => void;
  /**
   * 退出动画完成后的回调
   */
  onAfterClose?: () => void;
}

export interface UseAnimatedVisibleReturnType {
  transitionGroupProps: Omit<CSSTransitionProps, 'timeout'>;
  transitionGroupPropsWithoutListener: Required<Pick<CSSTransitionProps, 'in' | 'mountOnEnter' | 'unmountOnExit'>>;
  styleDisplay: React.CSSProperties;
  /**
   * 是否进入打开状态。进入动画开始后或退出动画结束前为 true
   */
  animatedVisible: boolean;
  /**
   * 是否进入打开稳态。进入动画完成或退出动画开始前为 true
   */
  enteredVisible: boolean;
  /**
   * 用于逻辑 Transition 组件的 ref
   */
  noopRef: React.MutableRefObject<HTMLElement>;
}

export const useTransitionGroupNoopNodeRef = () =>
  // https://github.com/reactjs/react-transition-group/blob/5aa3fd2d7e3354a7e42505d55af605ff44f74e2e/src/Transition.js#L325
  // 必须得给个 nodeRef
  // 1. 给 nodeRef 才能不 findDOMNode
  // 2. nodeRef 不能为空，不然会不等待 timeout
  React.useRef(__DEV__ && isClient() ? document.createElement('div') : ({} as HTMLElement));
export const useAnimatedVisible = ({
  forceRender = false,
  destroyOnClose = false,
  visible,
  onAfterOpen,
  onAfterClose,
}: UseAnimatedVisibleProps): UseAnimatedVisibleReturnType => {
  const transitionGroupPropsWithoutListener = React.useMemo(
    () => ({
      in: visible,
      mountOnEnter: !forceRender,
      unmountOnExit: destroyOnClose,
    }),
    [destroyOnClose, forceRender, visible],
  );

  const [animatedVisible, setAnimatedVisible] = React.useState<boolean>(visible);
  const [enteredVisible, setEnteredVisible] = React.useState<boolean>(false);

  const noopRef = useTransitionGroupNoopNodeRef();

  /**
   * 模拟 getDerivedStateFromProps，如果从 visible false -> true，第一次 render  animatedVisible 还是 false，所有 child mounted 的时候父元素还是 display none，这样一些计算就没有办法做，能让用户在 content mounted 的时候拿到正确的 display 是我们的 feature
   */
  const [prevVisible, setPrevVisible] = React.useState<boolean | null>(null);
  if (prevVisible !== visible) {
    if (visible) {
      setAnimatedVisible(true);
    }
    setPrevVisible(visible);
  }

  return {
    // 因为我们在 trigger 中会把这些值作为 context 传下去，为了 context selector 能帮我们做 render 性能优化，所以都用 memo 包裹一下
    noopRef,
    transitionGroupProps: React.useMemo(
      () => ({
        onEntered: () => {
          setEnteredVisible(true);
          isFunction(onAfterOpen) && onAfterOpen();
        },
        onExit: () => {
          setEnteredVisible(false);
        },
        onExited: () => {
          setAnimatedVisible(false);
          onAfterClose?.();
        },
        ...transitionGroupPropsWithoutListener,
      }),
      [onAfterClose, onAfterOpen, transitionGroupPropsWithoutListener],
    ),
    transitionGroupPropsWithoutListener,
    styleDisplay: React.useMemo(
      () =>
        animatedVisible
          ? {}
          : {
              display: 'none',
            },
      [animatedVisible],
    ),
    animatedVisible,
    enteredVisible,
  };
};
