// @universe-design/styles/esm/data-v/index.light.atom.js
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
export const datav_categorical_01 = '#3370eb';
export const datav_categorical_02 = '#1bcebf';
export const datav_categorical_03 = '#ffc60a';
export const datav_categorical_04 = '#ed6d0c';
export const datav_categorical_05 = '#dca1e4';
export const datav_categorical_06 = '#25b2e5';
export const datav_categorical_07 = '#6dcdeb';
export const datav_categorical_08 = '#288fcb';
export const datav_categorical_09 = '#94b5f5';
export const datav_categorical_10 = '#8f61d1';
export const datav_categorical_11 = '#bf78e9';
export const datav_categorical_12 = '#008280';
export const datav_categorical_13 = '#27ad8e';
export const datav_categorical_14 = '#7bc335';
export const datav_fill_hover = 'rgba(31, 35, 41, 0.05)';
export const datav_function_element_danger = '#ff615d';
export const datav_function_element_info = '#5083fb';
export const datav_function_element_neutral = '#b9bec6';
export const datav_function_element_success = '#47cfb8';
export const datav_function_element_warning = '#ff814a';
export const datav_function_text_danger = '#cd4841';
export const datav_function_text_info = '#2c5cc5';
export const datav_function_text_success = '#2d8639';
export const datav_function_text_warning = '#c75f1c';
export const datav_line_axis = 'rgba(31, 35, 41, 0.3)';
export const datav_line_default = 'rgba(31, 35, 41, 0.1)';
