// @universe-design/react/lib/message/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Toast", {
  enumerable: true,
  get: function () {
    return _Toast.Toast;
  }
});
exports.getGlobalToast = getGlobalToast;
exports.message = void 0;
exports.renderGlobalToast = renderGlobalToast;
exports.useMessage = exports.toast = void 0;
exports.useToast = useToast;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireWildcard(require("react"));
var _reactDom = _interopRequireDefault(require("react-dom"));
var _configProvider = require("../config-provider");
var _Toast = require("./Toast");
require("./style");
let globalToast;
let globalToastInitPromise;
function getDefaultContainer() {
  const div = document.createElement('div');
  document.body.appendChild(div);
  return div;
}
function singletonRef(elem) {
  if (elem) {
    globalToast = elem;
  }
}

/**
 * 用于多次重复 renderGlobalToast 和销毁自动生成的 DOM 元素
 */
class ToastWrapper extends _react.default.Component {
  constructor() {
    super(...arguments);
    (0, _defineProperty2.default)(this, "div", void 0);
    (0, _defineProperty2.default)(this, "isDefaultContainer", true);
    (0, _defineProperty2.default)(this, "getContainer", () => {
      if (!this.div) {
        var _Toast$globalConfig$g, _Toast$globalConfig;
        const globalContainer = (_Toast$globalConfig$g = (_Toast$globalConfig = _Toast.Toast.globalConfig).getContainer) === null || _Toast$globalConfig$g === void 0 ? void 0 : _Toast$globalConfig$g.call(_Toast$globalConfig);
        this.div = globalContainer || getDefaultContainer();
        if (globalContainer) {
          this.isDefaultContainer = false;
        }
      }
      return this.div;
    });
  }
  componentWillUnmount() {
    globalToast = undefined;
    if (this.isDefaultContainer && this.div) {
      this.div.remove();
      this.div = undefined;
    }
  }
  render() {
    return /*#__PURE__*/_react.default.createElement(_Toast.Toast, (0, _extends2.default)({}, this.props, {
      ref: singletonRef,
      getContainer: this.getContainer
    }));
  }
}
function getGlobalToast() {
  if (globalToast) {
    return Promise.resolve(globalToast);
  }
  if (globalToastInitPromise) {
    return globalToastInitPromise;
  }
  globalToastInitPromise = new Promise(resolve => {
    const div = getDefaultContainer();
    // 如果没有使用 useToast 或者 renderGlobalToast 初始化 toast，
    // 则默认取第一个 ConfigProvider 提供的配置
    _reactDom.default.render(/*#__PURE__*/_react.default.createElement(_configProvider.CrossRootConfigProvider, null, /*#__PURE__*/_react.default.createElement(ToastWrapper, {
      ref: () => {
        if (globalToast) {
          globalToastInitPromise = undefined;
          resolve(globalToast);
        }
      },
      getContainer: () => div
    })), div);
  });
  return globalToastInitPromise;
}
function createTostWrapper(getToast) {
  return {
    open(config) {
      return getToast().then(toast => toast.open(config) || '');
    },
    info(configOrContent, duration, onClose) {
      return getToast().then(toast => toast.info(configOrContent, duration, onClose));
    },
    success(configOrContent, duration, onClose) {
      return getToast().then(toast => toast.success(configOrContent, duration, onClose));
    },
    warning(configOrContent, duration, onClose) {
      return getToast().then(toast => toast.warning(configOrContent, duration, onClose));
    },
    warn(configOrContent, duration, onClose) {
      return getToast().then(toast => toast.warn(configOrContent, duration, onClose));
    },
    error(configOrContent, duration, onClose) {
      return getToast().then(toast => toast.error(configOrContent, duration, onClose));
    },
    loading(configOrContent, duration, onClose) {
      return getToast().then(toast => toast.loading(configOrContent, duration, onClose));
    },
    config(options) {
      _Toast.Toast.globalConfig = Object.assign(_Toast.Toast.globalConfig, options);
    },
    destroy(key) {
      getToast().then(toast => toast.destroy(key));
    }
  };
}
const toast = exports.toast = createTostWrapper(getGlobalToast);

/**
 * 用于创建可以获取 ConfigContext 的全局 Toast 单例
 * @example
 * ```tsx
 * <ConfigProvider>
 *   {renderGlobalToast({ className: 'my-toast' }, document.getElementById('my-container'))}
 *   <App />
 * </ConfigProvider>
 * ```
 *
 * @param props: ToastProps
 * @returns React.ReactPortal
 */
function renderGlobalToast(props) {
  return /*#__PURE__*/_react.default.createElement(ToastWrapper, props);
}

/**
 * 获取一个局部的 toast 实例
 * @returns
 */
function useToast(props, container) {
  const toastRef = (0, _react.useRef)(null);
  const mountDivRef = (0, _react.useRef)();
  const toastIns = createTostWrapper(() => {
    if (toastRef.current) {
      return Promise.resolve(toastRef.current);
    }
    return getGlobalToast();
  });
  const _getDefaultContainer = () => {
    if (mountDivRef.current) {
      return mountDivRef.current;
    }
    mountDivRef.current = getDefaultContainer();
    return mountDivRef.current;
  };
  return [toastIns, /*#__PURE__*/_react.default.createElement(_configProvider.ConfigProvider, {
    key: "ud-hook-toast-wrapper"
  }, /*#__PURE__*/_react.default.createElement(_Toast.Toast, (0, _extends2.default)({
    key: "ud-hook-toast"
  }, props, {
    ref: toastRef,
    getContainer: () => container || _getDefaultContainer()
  })))];
}

// 兼容竞品 API
const message = exports.message = toast;
const useMessage = exports.useMessage = useToast;
