// @universe-design/react/src/color-picker/components/ColorPickerSaturation.tsx
import React, { useRef } from 'react';
import clamp from 'lodash/clamp';
import clsx from 'clsx';
import { useColorPickerContextSelector } from '../ColorPickerContext';
import composeRefs from '../../util/composeRefs';
import { isFunction } from '../../util';
import { ColorPickerChangeSource } from '../type';
import { formatColor } from '../util';

/**
 * @public
 */
export interface ColorPickerSaturationProps extends React.ComponentPropsWithoutRef<'div'> {
  /**
   * 饱和度滑块拖动结束（pointerUp）
   */
  onSaturationDragEnd?: (e: PointerEvent) => void;
}

/**
 * 调整饱和度组件
 */
export const ColorPickerSaturation = React.forwardRef<HTMLDivElement, ColorPickerSaturationProps>(
  ({ className, style, onMouseDown, onPointerDown, onSaturationDragEnd, ...props }, ref) => {
    const setValue = useColorPickerContextSelector(o => o.setValue);
    const rootCssScope = useColorPickerContextSelector(o => o.rootCssScope);
    const displayColor = useColorPickerContextSelector(o => o.displayColor);
    const transformDisplayColor = useColorPickerContextSelector(o => o.transformDisplayColor);

    const hsv = useColorPickerContextSelector(o => o.hsv);
    const { hsv: displayHSV } = transformDisplayColor(formatColor(hsv));
    const prefix = rootCssScope('-saturation');
    const divRef = useRef<HTMLDivElement>(null);
    const mergedRef = React.useMemo(() => composeRefs(divRef, ref), [ref]);

    const position = {
      top: 100 - hsv.v * 100,
      left: hsv.s * 100,
    };

    const getPosition = (event: PointerEvent) => {
      const el = divRef.current;
      if (!el) {
        return position;
      }
      const containerRect = el.getBoundingClientRect();
      const { width: containerWidth, height: containerHeight } = containerRect;

      let left = event.pageX - (containerRect.left + window.pageXOffset);
      let top = event.pageY - (containerRect.top + window.pageYOffset);
      left = clamp(left, 0, containerWidth);
      top = clamp(top, 0, containerHeight);
      return {
        left: (left * 100) / containerWidth,
        top: (top * 100) / containerHeight,
      };
    };

    const handleChange = (event: PointerEvent) => {
      const pos = getPosition(event);
      setValue(
        {
          h: hsv.h,
          s: pos.left / 100,
          v: 1 - pos.top / 100,
          a: hsv.a,
        },
        { source: ColorPickerChangeSource.SATURATION },
      );
    };

    const handlePointerUp = (e: PointerEvent) => {
      onSaturationDragEnd?.(e);
      window.removeEventListener('pointermove', handleChange);
      window.removeEventListener('pointerup', handlePointerUp);
    };

    const bindEventListeners = () => {
      window.addEventListener('pointermove', handleChange);
      window.addEventListener('pointerup', handlePointerUp);
    };

    const handlePointerDown = (event: React.PointerEvent<HTMLDivElement>) => {
      isFunction(onMouseDown) && onMouseDown(event);
      isFunction(onPointerDown) && onPointerDown(event);
      handleChange(event.nativeEvent);
      bindEventListeners();
    };

    return (
      <div
        {...props}
        className={clsx(prefix, className)}
        style={{
          ...style,
          background: displayColor && `hsl(${displayHSV.h}, 100%, 50%)`,
        }}
        ref={mergedRef}
        onPointerDown={handlePointerDown}
      >
        <div className={`${prefix}-border`} />
        <div className={`${prefix}-white`}>
          <div className={`${prefix}-black`} />
          <div
            className={clsx(
              `${prefix}-circle`,
              // 暂时先去掉这个 circle 的 judge，全部显示白色就行，这个和 https://www.google.com.hk/search?q=color+picker 的行为一致
              // judgeNeedDeepenIcon(hex) && `${prefix}-circle-deepen`
            )}
            style={{
              top: `${position.top}%`,
              left: `${position.left}%`,
              background: displayColor && displayColor.rgbString,
            }}
          />
        </div>
      </div>
    );
  },
);
