// @universe-design/react/src/color-picker/components/ColorPickerHue.tsx
/** @jsxRuntime classic */
/** @jsx jsx */
import React from 'react';
import stylex from '@universe-design/stylex';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx } from '@universe-design/stylex/react';
import clsx from 'clsx';
import clamp from 'lodash/clamp';
import { useColorPickerContextSelector } from '../ColorPickerContext';
import { formatColor } from '../util';
import { ColorPickerChangeSource } from '../type';

export interface ColorPickerHueProps extends React.ComponentPropsWithoutRef<'div'> {
  /**
   * 色调滑块拖动结束（pointerUp）
   */
  onHueDragEnd?: (e: PointerEvent) => void;
}

export const ColorPickerHue = React.forwardRef<HTMLDivElement, ColorPickerHueProps>(
  ({ className, onHueDragEnd, ...props }, ref) => {
    const setValue = useColorPickerContextSelector(o => o.setValue);
    const rootCssScope = useColorPickerContextSelector(o => o.rootCssScope);
    const transformDisplayColor = useColorPickerContextSelector(o => o.transformDisplayColor);
    const displayColor = useColorPickerContextSelector(o => o.displayColor);
    const hsv = useColorPickerContextSelector(o => o.hsv);
    const prefix = rootCssScope('-hue');
    const sliderHandlerRef = React.useRef<HTMLDivElement>(null);

    const calculateChange = (event: PointerEvent) => {
      const el = sliderHandlerRef.current;
      if (!el) {
        return 0;
      }
      const containerWidth = el.clientWidth;
      const distance = event.pageX - (el.getBoundingClientRect().left + window.pageXOffset);
      const val = clamp(distance / containerWidth, 0, 1);
      return Math.round(val * 100);
    };

    const handleChange = (event: PointerEvent) => {
      const value = calculateChange(event);
      // eslint-disable-next-line @typescript-eslint/no-magic-numbers
      const h = clamp((value * 360) / 100, 0, 360);
      setValue(
        {
          ...hsv,
          h,
        },
        { source: ColorPickerChangeSource.HUE },
      );
    };

    const handlePointerUp = (e: PointerEvent) => {
      onHueDragEnd?.(e);
      window.removeEventListener('pointermove', handleChange);
      window.removeEventListener('pointerup', handlePointerUp);
    };

    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    const left = `${(hsv.h * 100) / 360}%`;

    const handlePointerDown = (event: React.PointerEvent<HTMLDivElement>) => {
      handleChange(event.nativeEvent);
      window.addEventListener('pointermove', handleChange);
      window.addEventListener('pointerup', handlePointerUp);
    };

    const sliderBackground = React.useMemo(
      () =>
        `linear-gradient(to right, ${transformDisplayColor(formatColor('#f00')).rgbString} 0%, ${
          transformDisplayColor(formatColor('#ff0')).rgbString
        } 17%, ${transformDisplayColor(formatColor('#0f0')).rgbString} 33%, ${
          transformDisplayColor(formatColor('#0ff')).rgbString
        } 50%, ${transformDisplayColor(formatColor('#00f')).rgbString} 67%, ${
          transformDisplayColor(formatColor('#f0f')).rgbString
        } 83%, ${transformDisplayColor(formatColor('#f00')).rgbString} 100%)`,
      [transformDisplayColor],
    );

    return (
      <div {...props} ref={ref} className={clsx(prefix, className)}>
        <div
          className={clsx(`${prefix}-display`, `${prefix}-display-highlight`)}
          style={{
            background: displayColor && displayColor.rgbString,
          }}
        />
        <div
          className={`${prefix}-slider`}
          ref={sliderHandlerRef}
          onPointerDown={handlePointerDown}
          style={{
            background: sliderBackground,
          }}
        >
          <div className={`${prefix}-pointer`} style={{ left }}>
            {/* /* 
             用 svg 做一个顶部的尖尖
             这个尖尖会带 1px 的底，和下方的长方形重叠
             防止缩放时，2 者出现间隙
             */}
            <svg
              width={12}
              height={5}
              sx={[
                stylex.create({
                  position: 'absolute',
                  top: -4,
                }),
              ]}
            >
              <path d="M0,4 V5 H10 V4 L5,0 Z" style={{ fill: 'currentcolor' }} />
            </svg>
          </div>
        </div>
      </div>
    );
  },
);
