// @universe-design/react/lib/carousel/Carousel.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Carousel = void 0;
var _u1jmni = _interopRequireDefault(require("./../.stylex/u1jmni06"));
var _u9yw1hi = _interopRequireDefault(require("./../.stylex/u9yw1hi"));
var _u1u1p7mj = _interopRequireDefault(require("./../.stylex/u1u1p7mj"));
var _us2xlyu = _interopRequireDefault(require("./../.stylex/us2xlyu"));
var _u184z4yl = _interopRequireDefault(require("./../.stylex/u184z4yl"));
var _u5j3c7v = _interopRequireDefault(require("./../.stylex/u5j3c7v"));
var _u8rtfz = _interopRequireDefault(require("./../.stylex/u8rtfz1"));
var _u1j6p86f = _interopRequireDefault(require("./../.stylex/u1j6p86f"));
var _ukchsq = _interopRequireDefault(require("./../.stylex/ukchsq7"));
var _uohwdt = _interopRequireDefault(require("./../.stylex/uohwdt"));
var _udg2fzc = _interopRequireDefault(require("./../.stylex/udg2fzc"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _clsx = _interopRequireDefault(require("clsx"));
var _CarouselContext = require("./CarouselContext");
var _CarouselWindow = require("./CarouselWindow");
var _CarouselSwitcher = require("./CarouselSwitcher");
var _CarouselIndicator = require("./CarouselIndicator");
var _CarouselManager = require("./CarouselManager");
var _defaults = require("./defaults");
var _fade = require("./transition-descriptors/fade");
var _push = require("./transition-descriptors/push");
//.u1ktbn1t{left:50%;}
//.u13gfj74{transform:translateX(-50%);}
//.u1euv43f{position:absolute;}
//.uf2sm71{height:48px;}
//.u1xynx9j{left:0px;}
//.u10k790i{right:0px;}
//.u5gq2j6{bottom:0px;}
//.ufsbof8{background-image:linear-gradient(0deg, @static-black-15 0%, transparent 80.21%);}
//.u22iagw{display:flex;}
//.u1vx9l62{flex-direction:column;}
//.u1l02sjl{height:100%;}
const ComposableCarousel = _ref => {
  let {
    children,
    className,
    style,
    ...useCarouselProps
  } = _ref;
  const {
    manager,
    contextValue
  } = (0, _CarouselContext.useCarousel)(useCarouselProps);
  const {
    rootPrefixCls
  } = contextValue;
  return /*#__PURE__*/React.createElement(_CarouselManager.CarouselManagerContextProvider, {
    value: manager
  }, /*#__PURE__*/React.createElement(_CarouselContext.CarouselContextProvider, {
    value: contextValue
  }, /*#__PURE__*/React.createElement("div", {
    className: (0, _clsx.default)(rootPrefixCls, (0, _stylex.default)({
      "props": {
        "mc9l5x": _ukchsq.default,
        "Beiy3e4": _uohwdt.default,
        "Bqenvij": _udg2fzc.default
      }
    }), className),
    style
  }, children)));
};
const CarouselMask = () => {
  const {
    rootPrefixCls
  } = (0, _CarouselContext.useCarouselContext)();
  return /*#__PURE__*/React.createElement("div", {
    className: (0, _clsx.default)(`${rootPrefixCls}__mask`, (0, _stylex.default)({
      "props": {
        "qhf8xq": _u1u1p7mj.default,
        "Bqenvij": _us2xlyu.default,
        "oyh7mz": _u184z4yl.default,
        "j35jbq": _u5j3c7v.default,
        "B5kzvoi": _u8rtfz.default,
        "Bcmaq0h": _u1j6p86f.default
      }
    }))
  });
};

/**
 * @public
 * @docgen
 */

/**
 * @public
 */
const Carousel = _ref2 => {
  let {
    children,
    indicatorProps,
    indicatorPosition = _defaults.carouselDefaults.indicatorPosition,
    switcherProps,
    transition = _defaults.carouselDefaults.transition,
    ...sharedProps
  } = _ref2;
  return /*#__PURE__*/React.createElement(ComposableCarousel, (0, _extends2.default)({}, sharedProps, {
    slides: children,
    transitionDescriptor: (() => {
      if (transition === 'fade') {
        return new _fade.FadeCarouselTransitionDescriptor();
      }
      return new _push.PushCarouselTransitionDescriptor();
    })()
  }), /*#__PURE__*/React.createElement(_CarouselWindow.CarouselWindow, null, /*#__PURE__*/React.createElement(_CarouselSwitcher.CarouselPrevSwitcher, switcherProps), /*#__PURE__*/React.createElement(_CarouselSwitcher.CarouselNextSwitcher, switcherProps), indicatorPosition === 'inner-bottom' && /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(CarouselMask, null), /*#__PURE__*/React.createElement(_CarouselIndicator.CarouselIndicator, (0, _extends2.default)({
    theme: "light"
  }, indicatorProps, {
    sx: [{
      "props": {
        "qhf8xq": _u1u1p7mj.default,
        "B5kzvoi": _u8rtfz.default,
        "oyh7mz": _u1jmni.default,
        "Bz10aip": _u9yw1hi.default
      }
    }, indicatorProps === null || indicatorProps === void 0 ? void 0 : indicatorProps.sx]
  })))), indicatorPosition === 'outer-bottom' && /*#__PURE__*/React.createElement(_CarouselIndicator.CarouselIndicator, (0, _extends2.default)({
    theme: "primary"
  }, indicatorProps)));
};
exports.Carousel = Carousel;
