// @universe-design/react/lib/carousel/CarouselSlide.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useCarouselSlideContext = exports.CarouselSlideContextProvider = exports.CarouselSlide = void 0;
var _udg2fzc = _interopRequireDefault(require("./../.stylex/udg2fzc"));
var _u13d3rh = _interopRequireDefault(require("./../.stylex/u13d3rh8"));
var _u1u1p7mj = _interopRequireDefault(require("./../.stylex/u1u1p7mj"));
var _u184z4yl = _interopRequireDefault(require("./../.stylex/u184z4yl"));
var _u1981n0b = _interopRequireDefault(require("./../.stylex/u1981n0b"));
var _u8wgnjj = _interopRequireDefault(require("./../.stylex/u8wgnjj"));
var _u1p85sbs = _interopRequireDefault(require("./../.stylex/u1p85sbs"));
var _u19qsyxj = _interopRequireDefault(require("./../.stylex/u19qsyxj"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var React = _interopRequireWildcard(require("react"));
var _reactTransitionGroup = require("react-transition-group");
var _clsx = _interopRequireDefault(require("clsx"));
var _CarouselContext = require("./CarouselContext");
var _CarouselManager = require("./CarouselManager");
var _reactUtilities = require("@universe-design/react-utilities");
var _hooks = require("../hooks");
var _composeRefs = _interopRequireDefault(require("../util/composeRefs"));
var _forwardRef = require("../util/forwardRef");
var _pick = _interopRequireDefault(require("lodash/pick"));
var _omit = _interopRequireDefault(require("lodash/omit"));
//.u1l02sjl{height:100%;}
//.uly5x3f{width:100%;}
//.u1euv43f{position:absolute;}
//.u1xynx9j{left:0px;}
//.u113wtx2{top:0px;}
//.u14t3ns0{display:inline-block;}
//.u1ewtqcl{box-sizing:border-box;}
//.uq62nq6{background-color:@text-disabled;}
const transitionGroupClonePropKeys = ['onExited', 'in', 'appear', 'enter', 'exit'];
const {
  Provider: CarouselSlideContextProvider,
  useContext: useCarouselSlideContext
} = (0, _reactUtilities.createContextHook)('CarouselSlideContext', {
  strict: true
});
exports.useCarouselSlideContext = useCarouselSlideContext;
exports.CarouselSlideContextProvider = CarouselSlideContextProvider;
/**
 * @public
 */
const CarouselSlide = exports.CarouselSlide = (0, _forwardRef.forwardHTMLRef)((props, ref) => {
  const {
    children,
    className,
    fullHeight = true,
    fullWidth = true,
    ...restProps
  } = props;
  // https://github.com/reactjs/react-transition-group/blob/0d8a515aa59b8ba77fcabaac479e798532d4adfa/src/utils/ChildMapping.js#L88
  const transitionGroupProps = (0, _pick.default)(restProps, transitionGroupClonePropKeys);
  const domAttrs = (0, _omit.default)(restProps, transitionGroupClonePropKeys);
  const {
    groupIndex,
    slideIndex
  } = useCarouselSlideContext();
  const {
    duration,
    rootPrefixCls
  } = (0, _CarouselContext.useCarouselContext)();
  const manager = (0, _CarouselManager.useCarouselManagerContext)();
  const slideElRef = React.useRef(null);
  const slideElResizeObserverRef = (0, _reactUtilities.useResizeObserver)(entry => {
    const rect = (0, _hooks.getPositionAndSize)(entry.target, entry);
    // logger.info(
    //   `Carousel useResizeObserver ${JSON.stringify({
    //     groupIndex,
    //     slideIndex,
    //   })}`,
    //   {
    //     width: rect.width,
    //     height: rect.height,
    //   }
    // );
    manager.updateSlideSize({
      groupIndex,
      slideIndex
    }, rect);
  });
  const composedSlideElRef = React.useMemo(() => (0, _composeRefs.default)(ref, slideElRef, slideElResizeObserverRef, manager.register({
    groupIndex,
    slideIndex
  })), [groupIndex, manager, ref, slideElResizeObserverRef, slideIndex]);
  (0, _hooks.useIsomorphicLayoutEffect)(() => {
    // 首次渲染立即存下 slide size
    const positionedEl = slideElRef.current;
    if (!positionedEl) {
      return;
    }
    // force reflow 但是没关系，因为都是 absolute 的元素
    manager.updateSlideSize({
      groupIndex,
      slideIndex
    }, (0, _hooks.getPositionAndSize)(positionedEl));
  }, [groupIndex, manager, slideIndex]);

  // useLogger(`CarouselSlide ${JSON.stringify({ groupIndex, slideIndex })}`);

  return /*#__PURE__*/React.createElement(_reactTransitionGroup.Transition, (0, _extends2.default)({
    nodeRef: slideElRef,
    timeout: {
      exit: duration
    }
  }, transitionGroupProps), /*#__PURE__*/React.createElement("div", (0, _extends2.default)({
    ref: composedSlideElRef,
    "data-group-index": groupIndex,
    "data-slide-index": slideIndex,
    className: (0, _clsx.default)(`${rootPrefixCls}__slide`, (0, _stylex.default)({
      "props": {
        "qhf8xq": _u1u1p7mj.default,
        "oyh7mz": _u184z4yl.default,
        "Bhzewxz": _u1981n0b.default,
        "mc9l5x": _u8wgnjj.default,
        "B7ck84d": _u1p85sbs.default,
        "De3pzq": _u19qsyxj.default
      }
    }, fullWidth && {
      "props": {
        "a9b677": _u13d3rh.default
      }
    }, fullHeight && {
      "props": {
        "Bqenvij": _udg2fzc.default
      }
    }), className)
  }, domAttrs), children));
});

/**
 * @public
 * @docgen
 */
