// @universe-design/react/lib/carousel/CarouselWindow.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CarouselWindow = void 0;
var _ux645fm = _interopRequireDefault(require("./../.stylex/ux645fm"));
var _uugkpzq = _interopRequireDefault(require("./../.stylex/uugkpzq"));
var _u5jxv = _interopRequireDefault(require("./../.stylex/u5jxv71"));
var _u99gbyh = _interopRequireDefault(require("./../.stylex/u99gbyh"));
var React = _interopRequireWildcard(require("react"));
var _reactTransitionGroup = require("react-transition-group");
var _shim = require("use-sync-external-store/shim");
var _clsx = _interopRequireDefault(require("clsx"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _hooks = require("../hooks");
var _reactUtilities = require("@universe-design/react-utilities");
var _CarouselManager = require("./CarouselManager");
var _CarouselContext = require("./CarouselContext");
var _CarouselSlide = require("./CarouselSlide");
//.u1a3p1vp{overflow:hidden;}
//.uz5stix{white-space:nowrap;}
//.u10pi13n{position:relative;}
//.uqerorx{flex-grow:1;}
const CarouselSlideProvider = _ref => {
  let {
    children,
    slide,
    ...props
  } = _ref;
  return (
    /*#__PURE__*/
    // 这里用 ForwardProps 是因为 TransitionGroup 会给子元素 cloneProps，但是我们又需要给 list item 加 key
    React.createElement(_CarouselSlide.CarouselSlideContextProvider, {
      value: slide
    }, /*#__PURE__*/React.cloneElement(children, props))
  );
};
const CarouselWindow = _ref2 => {
  let {
    children
  } = _ref2;
  const {
    rootPrefixCls,
    duration,
    slide,
    slides,
    play,
    pause,
    autoPlayPauseOnHover
  } = (0, _CarouselContext.useCarouselContext)();
  const manager = (0, _CarouselManager.useCarouselManagerContext)();

  /**
   * ========================================
   *   Subscribe state changes
   * ========================================
   */
  const slidesToBeRendered = (0, _shim.useSyncExternalStore)(manager.subscribe, React.useCallback(() => manager.state.slidesToBeRendered, [manager]));
  const slideSizes = (0, _shim.useSyncExternalStore)(manager.subscribe, React.useCallback(() => manager.state.slideSizes, [manager]));
  const windowSize = (0, _shim.useSyncExternalStore)(manager.subscribe, React.useCallback(() => manager.state.windowSize, [manager]));

  /**
   * ========================================
   *   Window Size
   * ========================================
   */
  const ref = (0, _reactUtilities.useResizeObserver)(entry => {
    const rect = (0, _hooks.getPositionAndSize)(entry.target, entry);
    manager.updateWindowSize(rect);
  });
  (0, _hooks.useIsomorphicLayoutEffect)(() => {
    // 首次渲染
    const el = ref.current;
    if (!el) {
      return;
    }
    manager.updateWindowSize((0, _hooks.getPositionAndSize)(el));
  }, [manager, ref]);

  /**
   * ========================================
   *   Update Slides To Be Rendered
   * ========================================
   */
  (0, _hooks.useIsomorphicLayoutEffect)(() => {
    manager.updateSlidesToBeRendered();
  });

  /**
   * ========================================
   *  Reset All Position Start
   * ========================================
   * TODO 这段逻辑很不优雅
   *
   * - slideIndex 变化以后需要重新做动画（中间几张图比较小，切换 slideIndex 后 slidesToBeRendered 没有变的情况）
   * - windowSize 和 slideSizes 的变化，当不引起 slidesToBeRendered 变化时也需要一步到位（视口大小变化，幻灯片尺寸变化，但是未引起 slidesToBeRendered 时）
   * - slidesToBeRendered 变化有两种情况
   *  - slideIndex 变化导致的 slidesToBeRendered 变化，prevSlides 和 nextSlides 需要一起做动画
   *  - windowSize 和 slideSizes 导致的 slidesToBeRendered 变化，需要立即到位
   *
   */
  const changed = React.useRef({
    slide: false,
    windowSize: false,
    slideSizes: false
  });
  (0, _hooks.useIsomorphicLayoutEffect)(() => {
    // 这个逻辑是保证 slidesToBeRendered windowSize 变化也能触发直接跳转

    changed.current.windowSize = true;
    manager.resetNodes({
      slick: false
    });
  }, [manager, windowSize]);
  (0, _hooks.useIsomorphicLayoutEffect)(() => {
    // 这个逻辑是保证 slidesToBeRendered 不变时，slideSizes 变化也能触发直接跳转

    changed.current.slideSizes = true;
    manager.resetNodes({
      slick: false
    });
  }, [manager, slideSizes]);
  (0, _hooks.useIsomorphicLayoutEffect)(() => {
    // 这个逻辑是保证 slidesToBeRendered 不变时，slide 变化也能触发动画跳转
    const el = ref.current;
    if (!el) {
      return;
    }
    slide.groupIndex;
    slide.slideIndex;
    changed.current.slide = true;
    // 这里所有的 size 都已经稳定，所有 slides 都已经存在，重置一次位置，并开始做动画
    manager.resetNodes({
      slick: true
    });
  }, [duration, manager, ref, slide.groupIndex, slide.slideIndex]);
  (0, _hooks.useIsomorphicLayoutEffect)(() => {
    // 由于 updateSlidesToBeRendered 是在上面几个 change 过程的下一个 render 周期才会做出反应，所以需要判断下 slidesToBeRendered 变化是什么导致的，并做出对应的措施。
    const el = ref.current;
    if (!el) {
      return;
    }
    if (changed.current.slide) {
      changed.current.slide = false;
      changed.current.windowSize = false;
      changed.current.slideSizes = false;
      manager.resetNodes({
        slick: true
      });
    }
    if (changed.current.windowSize || changed.current.slideSizes) {
      changed.current.windowSize = false;
      changed.current.slideSizes = false;
      manager.resetNodes({
        slick: false
      });
    }
  }, [duration, manager, ref, slidesToBeRendered]);
  /**
   * ========================================
   *  Reset All Position End
   * ========================================
   */

  // useLogger('CarouselWindow', { slidesToBeRendered, slideSizes, slide });

  return /*#__PURE__*/React.createElement("div", {
    ref: ref,
    className: (0, _clsx.default)(`${rootPrefixCls}__window`, (0, _stylex.default)({
      "props": {
        "Bpg54ce": _ux645fm.default,
        "Huce71": _uugkpzq.default,
        "qhf8xq": _u5jxv.default,
        "Bh6795r": _u99gbyh.default
      }
    })),
    onMouseEnter: () => {
      if (!autoPlayPauseOnHover) {
        return;
      }
      pause();
    },
    onMouseLeave: () => {
      play();
    }
  }, /*#__PURE__*/React.createElement(_reactTransitionGroup.TransitionGroup, {
    className: `${rootPrefixCls}__track`
  }, slidesToBeRendered.map(_slide => /*#__PURE__*/React.createElement(CarouselSlideProvider, {
    key: `${_slide.groupIndex}-${_slide.slideIndex}`,
    slide: _slide
  }, slides[_slide.slideIndex]))), children);
};
exports.CarouselWindow = CarouselWindow;
