// @universe-design/react/lib/carousel/CarouselSwitcher.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CarouselPrevSwitcher = exports.CarouselNextSwitcher = void 0;
var _u1aqot9l = _interopRequireDefault(require("./../.stylex/u1aqot9l"));
var _u1u1p7mj = _interopRequireDefault(require("./../.stylex/u1u1p7mj"));
var _u1utdkg = _interopRequireDefault(require("./../.stylex/u1utdkg1"));
var _ufibavv = _interopRequireDefault(require("./../.stylex/ufibavv"));
var _ulanhzm = _interopRequireDefault(require("./../.stylex/ulanhzm"));
var _uwf09ar = _interopRequireDefault(require("./../.stylex/uwf09ar"));
var _ubr2xhg = _interopRequireDefault(require("./../.stylex/ubr2xhg"));
var _u1o367ev = _interopRequireDefault(require("./../.stylex/u1o367ev"));
var _uggvdc = _interopRequireDefault(require("./../.stylex/uggvdc4"));
var _un6nsv = _interopRequireDefault(require("./../.stylex/un6nsv3"));
var _u19bueuo = _interopRequireDefault(require("./../.stylex/u19bueuo"));
var _ukchsq = _interopRequireDefault(require("./../.stylex/ukchsq7"));
var _ubyhozq = _interopRequireDefault(require("./../.stylex/ubyhozq"));
var _ut7q8es = _interopRequireDefault(require("./../.stylex/ut7q8es"));
var _uathu2q = _interopRequireDefault(require("./../.stylex/uathu2q"));
var _u5f56vu = _interopRequireDefault(require("./../.stylex/u5f56vu"));
var _ugowvl = _interopRequireDefault(require("./../.stylex/ugowvl5"));
var _u1h28k4f = _interopRequireDefault(require("./../.stylex/u1h28k4f"));
var _u1qg4c = _interopRequireDefault(require("./../.stylex/u1qg4c09"));
var _ugmgj = _interopRequireDefault(require("./../.stylex/ugmgj40"));
var _u12o6wmt = _interopRequireDefault(require("./../.stylex/u12o6wmt"));
var _u5kb0jy = _interopRequireDefault(require("./../.stylex/u5kb0jy"));
var _up = _interopRequireDefault(require("./../.stylex/up10246"));
var _uit90k = _interopRequireDefault(require("./../.stylex/uit90k4"));
var _u5nolws = _interopRequireDefault(require("./../.stylex/u5nolws"));
var _u1ul1s2h = _interopRequireDefault(require("./../.stylex/u1ul1s2h"));
var _u15mrnzb = _interopRequireDefault(require("./../.stylex/u15mrnzb"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _LeftOutlined = _interopRequireDefault(require("@universe-design/icons-react/LeftOutlined"));
var _RightOutlined = _interopRequireDefault(require("@universe-design/icons-react/RightOutlined"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _clsx = _interopRequireDefault(require("clsx"));
var React = _interopRequireWildcard(require("react"));
var _stylex2 = require("../stylex");
var _forwardRef = require("../util/forwardRef");
var _CarouselContext = require("./CarouselContext");
var _CarouselManager = require("./CarouselManager");
//.ucq9r29{right:24px;}
//.u1euv43f{position:absolute;}
//.u1ycc5zk{left:24px;}
//.u1i1t8d1{top:50%;}
//.u188r07x{transform:translateY(-50%);}
//.ueqmc2u{width:40px;}
//.ubhnoac{height:40px;}
//.u4ybsrx{font-size:16px;}
//.u1w9dchk{width:28px;}
//.uxldao9{height:28px;}
//.uses1vf{font-size:14px;}
//.u22iagw{display:flex;}
//.u122n59{align-items:center;}
//.u4d9j23{justify-content:center;}
//.u19ibtj8{color:@static-white;}
//.u1c04qm4{background-color:@static-black-50;}
//.u1ern45e{border-top-style:none;}
//.u1n71otn{border-right-style:none;}
//.u1h8hb77{border-bottom-style:none;}
//.u1deefiw{border-left-style:none;}
//.u1k6fduh{cursor:pointer;}
//.u1mf1pf4{transition-property:background-color,color;}
//.ubf55e9{transition-duration:0.1s,0.1s;}
//.u6mlncw:hover{background-color:@static-black-80;}
//.uxdupyk:active{background-color:@static-black-80;}
//.u2hkw1w:focus-visible{outline-style:none;}
//.u17ot7lm:focus-visible{background-color:@static-black-80;}
/**
 * TODO: 等 button 迁移 stylex 之后用 ud button 改写
 */
const CarouselSwitcher = (0, _forwardRef.forwardHTMLRef)((_ref, ref) => {
  let {
    // omit
    children,
    // props
    className,
    sx,
    icon,
    size = 'medium',
    ...buttonProps
  } = _ref;
  return /*#__PURE__*/React.createElement("button", (0, _extends2.default)({
    ref: ref,
    type: "button",
    className: (0, _clsx.default)(className, (0, _stylex.default)({
      "props": {
        "mc9l5x": _ukchsq.default,
        "Bt984gj": _ubyhozq.default,
        "Brf1p80": _ut7q8es.default,
        "sj55zd": _uathu2q.default,
        "De3pzq": _u5f56vu.default,
        "icvyot": _ugowvl.default,
        "vrafjx": _u1h28k4f.default,
        "oivjwe": _u1qg4c.default,
        "wvpqe5": _ugmgj.default,
        "Bceei9c": _u12o6wmt.default,
        "Bmy1vo4": _u5kb0jy.default,
        "B3o57yi": _up.default,
        "Jwef8y": _uit90k.default,
        "ecr2s2": _u5nolws.default,
        "B486eqv": _u1ul1s2h.default,
        "xyvl0c": _u15mrnzb.default
      }
    }, _stylex2.sxBorderRadius.xl, _stylex2.sxTransitionTimingFunction.standard, {
      small: {
        "props": {
          "a9b677": _uggvdc.default,
          "Bqenvij": _un6nsv.default,
          "Be2twd7": _u19bueuo.default
        }
      },
      medium: {
        "props": {
          "a9b677": _uwf09ar.default,
          "Bqenvij": _ubr2xhg.default,
          "Be2twd7": _u1o367ev.default
        }
      }
    }[size], sx))
  }, buttonProps), icon);
});

/**
 * @public
 * @docgen
 */

const CarouselPrevSwitcher = exports.CarouselPrevSwitcher = (0, _forwardRef.forwardHTMLRef)((_ref2, ref) => {
  let {
    // omit
    children,
    // props
    onClick,
    sx,
    ...buttonProps
  } = _ref2;
  const {
    rootPrefixCls
  } = (0, _CarouselContext.useCarouselContext)();
  const manager = (0, _CarouselManager.useCarouselManagerContext)();
  return /*#__PURE__*/React.createElement(CarouselSwitcher, (0, _extends2.default)({
    ref: ref,
    icon: /*#__PURE__*/React.createElement(_LeftOutlined.default, null),
    className: `${rootPrefixCls}__prev-switcher`
  }, buttonProps, {
    sx: [{
      "props": {
        "qhf8xq": _u1u1p7mj.default,
        "oyh7mz": _u1utdkg.default,
        "Bhzewxz": _ufibavv.default,
        "Bz10aip": _ulanhzm.default
      }
    }, sx],
    onClick: ev => {
      onClick === null || onClick === void 0 ? void 0 : onClick(ev);
      if (!manager.canUserInteract) {
        return;
      }
      manager.setSlideIndexByStep(-1);
    }
  }));
});
const CarouselNextSwitcher = exports.CarouselNextSwitcher = (0, _forwardRef.forwardHTMLRef)((_ref3, ref) => {
  let {
    // omit
    children,
    // props
    onClick,
    sx,
    ...buttonProps
  } = _ref3;
  const {
    rootPrefixCls
  } = (0, _CarouselContext.useCarouselContext)();
  const manager = (0, _CarouselManager.useCarouselManagerContext)();
  return /*#__PURE__*/React.createElement(CarouselSwitcher, (0, _extends2.default)({
    ref: ref,
    icon: /*#__PURE__*/React.createElement(_RightOutlined.default, null),
    className: `${rootPrefixCls}__next-switcher`
  }, buttonProps, {
    sx: [{
      "props": {
        "qhf8xq": _u1u1p7mj.default,
        "j35jbq": _u1aqot9l.default,
        "Bhzewxz": _ufibavv.default,
        "Bz10aip": _ulanhzm.default
      }
    }],
    onClick: ev => {
      onClick === null || onClick === void 0 ? void 0 : onClick(ev);
      if (!manager.canUserInteract) {
        return;
      }
      manager.setSlideIndexByStep(1);
    }
  }));
});
