// @universe-design/react/lib/carousel/CarouselIndicator.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CarouselIndicator = void 0;
var _u1u1p7mj = _interopRequireDefault(require("./../.stylex/u1u1p7mj"));
var _u14zp1q = _interopRequireDefault(require("./../.stylex/u14zp1q4"));
var _u1ska7bq = _interopRequireDefault(require("./../.stylex/u1ska7bq"));
var _umr338j = _interopRequireDefault(require("./../.stylex/umr338j"));
var _u17syspr = _interopRequireDefault(require("./../.stylex/u17syspr"));
var _u165j2wq = _interopRequireDefault(require("./../.stylex/u165j2wq"));
var _uvfwkig = _interopRequireDefault(require("./../.stylex/uvfwkig"));
var _ud7fs9g = _interopRequireDefault(require("./../.stylex/ud7fs9g"));
var _u1d4y5un = _interopRequireDefault(require("./../.stylex/u1d4y5un"));
var _uka18dz = _interopRequireDefault(require("./../.stylex/uka18dz"));
var _uymufub = _interopRequireDefault(require("./../.stylex/uymufub"));
var _u9dhz2c = _interopRequireDefault(require("./../.stylex/u9dhz2c"));
var _u6rgktp = _interopRequireDefault(require("./../.stylex/u6rgktp"));
var _ub4ucl = _interopRequireDefault(require("./../.stylex/ub4ucl8"));
var _uq3qnqa = _interopRequireDefault(require("./../.stylex/uq3qnqa"));
var _u15s3zb = _interopRequireDefault(require("./../.stylex/u15s3zb6"));
var _u5jxv = _interopRequireDefault(require("./../.stylex/u5jxv71"));
var _u18eu6n = _interopRequireDefault(require("./../.stylex/u18eu6n2"));
var _u12o6wmt = _interopRequireDefault(require("./../.stylex/u12o6wmt"));
var _ut7q8es = _interopRequireDefault(require("./../.stylex/ut7q8es"));
var _u1xft5cu = _interopRequireDefault(require("./../.stylex/u1xft5cu"));
var _u11atsnw = _interopRequireDefault(require("./../.stylex/u11atsnw"));
var _u1co87pz = _interopRequireDefault(require("./../.stylex/u1co87pz"));
var _uhzi0ua = _interopRequireDefault(require("./../.stylex/uhzi0ua"));
var _uidkx = _interopRequireDefault(require("./../.stylex/uidkx17"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _forwardRef = require("../util/forwardRef");
var _CarouselContext = require("./CarouselContext");
var _stylex2 = require("../stylex");
var _space = require("../space");
var _clsx = _interopRequireDefault(require("clsx"));
var _CarouselManager = require("./CarouselManager");
//.u1euv43f{position:absolute;}
//.uclvdme{left:-3px;}
//.uvq7pei{right:-3px;}
//.u12v7yn{top:-3px;}
//.uode0jj{bottom:-3px;}
//.u16dn6v3{width:6px;}
//.uqecqfw{background-color:@static-white-30;}
//.u12w28jm:hover{background-color:@static-white-60;}
//.u1qhvef2{background-color:@line-border-component;}
//.u1qh4al8:hover{background-color:@primary-pri-200;}
//.uq4mcun{width:24px;}
//.u6ywr7j{height:4px;}
//.urx94fk{width:12px;}
//.u3mu39s{height:6px;}
//.u1su79po{background-color:@primary-on-primary-fill;}
//.u1oquvwe{background-color:@primary-fill-default;}
//.u10pi13n{position:relative;}
//.u7o03ib{transition-property:all;}
//.u1k6fduh{cursor:pointer;}
//.u4d9j23{justify-content:center;}
//.uc7z3ec{padding-top:20px;}
//.uhxju0i{padding-right:0;}
//.ue2my4m{padding-bottom:20px;}
//.u1cnd47f{padding-left:0;}
//.ui64zpg{flex-shrink:0;}
/**
 * @public
 * @docgen
 */
const CarouselIndicator = exports.CarouselIndicator = (0, _forwardRef.forwardHTMLRef)((props, ref) => {
  const {
    className,
    type = 'dot',
    theme = 'light',
    sx,
    ...restProps
  } = props;
  const {
    slide,
    slideCount,
    duration,
    rootPrefixCls
  } = (0, _CarouselContext.useCarouselContext)();
  const manager = (0, _CarouselManager.useCarouselManagerContext)();
  return /*#__PURE__*/React.createElement(_space.Space, (0, _extends2.default)({
    className: (0, _clsx.default)(`${rootPrefixCls}__indicator`, (0, _stylex.default)({
      "props": {
        "Brf1p80": _ut7q8es.default,
        "z8tnut": _u1xft5cu.default,
        "z189sj": _u11atsnw.default,
        "Byoj8tv": _u1co87pz.default,
        "uwmqm3": _uhzi0ua.default,
        "Bnnss6s": _uidkx.default
      }
    }, sx), className),
    size: 6,
    ref: ref,
    inline: false
  }, restProps), Array(slideCount).fill(0).map((_, i) => /*#__PURE__*/React.createElement("div", {
    key: i,
    style: {
      transitionDuration: `${duration}ms`
    },
    className: (0, _clsx.default)(`${rootPrefixCls}__indicator-item`, `${rootPrefixCls}__indicator-item-${theme}`, `${rootPrefixCls}__indicator-item-${type}`, slide.slideIndex === i && `${rootPrefixCls}__indicator-item-active`, (0, _stylex.default)({
      "props": {
        "qhf8xq": _u5jxv.default,
        "Bmy1vo4": _u18eu6n.default,
        "Bceei9c": _u12o6wmt.default
      }
    }, _stylex2.sxBorderRadius.xl, _stylex2.sxTransitionTimingFunction.standard, slide.slideIndex === i ? [theme === 'primary' ? {
      "props": {
        "De3pzq": _u15s3zb.default
      }
    } : {
      "props": {
        "De3pzq": _uq3qnqa.default
      }
    }, type === 'dot' ? {
      "props": {
        "a9b677": _u6rgktp.default,
        "Bqenvij": _ub4ucl.default
      }
    } : {
      "props": {
        "a9b677": _uymufub.default,
        "Bqenvij": _u9dhz2c.default
      }
    }] : [theme === 'primary' ? {
      "props": {
        "De3pzq": _u1d4y5un.default,
        "Jwef8y": _uka18dz.default
      }
    } : {
      "props": {
        "De3pzq": _uvfwkig.default,
        "Jwef8y": _ud7fs9g.default
      }
    }, type === 'dot' ? {
      "props": {
        "a9b677": _u165j2wq.default,
        "Bqenvij": _ub4ucl.default
      }
    } : {
      "props": {
        "a9b677": _uymufub.default,
        "Bqenvij": _u9dhz2c.default
      }
    }])),
    onClick: () => {
      if (!manager.canUserInteract) {
        return;
      }
      manager.setSlideIndex(i);
    }
  }, /*#__PURE__*/React.createElement("div", {
    className: (0, _stylex.default)({
      "props": {
        "qhf8xq": _u1u1p7mj.default,
        "oyh7mz": _u14zp1q.default,
        "j35jbq": _u1ska7bq.default,
        "Bhzewxz": _umr338j.default,
        "B5kzvoi": _u17syspr.default
      }
    })
  }))));
});
