// @universe-design/react/lib/carousel/CarouselContext.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useCarouselContext = exports.useCarousel = exports.CarouselContextProvider = void 0;
var React = _interopRequireWildcard(require("react"));
var _reactUtilities = require("@universe-design/react-utilities");
var _clamp = _interopRequireDefault(require("lodash/clamp"));
var _shim = require("use-sync-external-store/shim");
var _defaults = require("./defaults");
var _configProvider = require("../config-provider");
var _util = require("../util");
var _CarouselManager = require("./CarouselManager");
/**
 * ====================
 *   AutoPlay
 * ====================
 */

/**
 * @public
 */

const getNow = () => new Date().getTime();

/**
 * @public
 */

const useCarouselAutoPlay = (_ref, _ref2) => {
  let {
    autoPlay = _defaults.carouselDefaults.autoPlay
  } = _ref;
  let {
    manager,
    slide
  } = _ref2;
  const {
    interval: autoPlayInterval,
    pauseOnHover: autoPlayPauseOnHover
  } = ((_autoPlay$interval, _autoPlay$pauseOnHove) => {
    if (!autoPlay || autoPlay === true) {
      return _defaults.carouselDefaults.autoPlayConfig;
    }
    return {
      interval: (_autoPlay$interval = autoPlay.interval) !== null && _autoPlay$interval !== void 0 ? _autoPlay$interval : _defaults.carouselDefaults.autoPlayConfig.interval,
      pauseOnHover: (_autoPlay$pauseOnHove = autoPlay.pauseOnHover) !== null && _autoPlay$pauseOnHove !== void 0 ? _autoPlay$pauseOnHove : _defaults.carouselDefaults.autoPlayConfig.pauseOnHover
    };
  })();
  const autoPlayEnabled = Boolean(autoPlay);
  const [isAutoPlayRunning, setIsAutoPlayRunning, isAutoPlayRunningRef] = (0, _reactUtilities.useLatestState)(false);
  // 切换幻灯片的时间
  const slideChangedAt = React.useRef(getNow());
  const playedAt = React.useRef(getNow());
  const pausedAt = React.useRef(getNow());

  // play ----> pause ----[idle]----> play ----> slide changed -[animation]->-------> slide changed ...
  React.useEffect(() => {
    // slide 变化的时候需要更新所有的时间
    slide.groupIndex;
    slide.slideIndex;
    const now = getNow();
    slideChangedAt.current = now;
    playedAt.current = now;
    pausedAt.current = now;
  }, [slide.groupIndex, slide.slideIndex]);
  React.useEffect(() => {
    // 开启 autoPlay 并且非暂停状态
    if (!autoPlayEnabled || !isAutoPlayRunning) {
      return;
    }
    let t;
    const loop = () => {
      t = window.setTimeout(() => {
        const elapsed = pausedAt.current - slideChangedAt.current + (getNow() - playedAt.current);
        if (elapsed > autoPlayInterval) {
          manager.setSlideIndexByStep(1);
        }
        loop();
      }, 100);
    };
    loop();
    return () => {
      window.clearTimeout(t);
    };
  }, [autoPlayEnabled, autoPlayInterval, isAutoPlayRunning, manager]);
  const play = React.useCallback(() => {
    if (isAutoPlayRunningRef.current) {
      return;
    }
    playedAt.current = getNow();
    setIsAutoPlayRunning(true);
  }, [isAutoPlayRunningRef, setIsAutoPlayRunning]);
  const pause = React.useCallback(() => {
    if (!isAutoPlayRunningRef.current) {
      return;
    }
    pausedAt.current = getNow();
    setIsAutoPlayRunning(false);
  }, [isAutoPlayRunningRef, setIsAutoPlayRunning]);
  React.useEffect(() => {
    if (autoPlay) {
      play();
    }
  }, [autoPlay, play]);
  return {
    play,
    pause,
    autoPlayPauseOnHover
  };
};

/**
 * @public
 */

/**
 * Carousel hook
 */
const useCarousel = props => {
  const {
    slides = _defaults.carouselDefaults.slides,
    defaultSlideIndex = _defaults.carouselDefaults.defaultSlideIndex,
    slideIndex: slideIndexProp,
    // infinite = carouselDefaults.infinite,
    onSlideIndexChange,
    autoPlay,
    transitionDescriptor,
    duration = transitionDescriptor.defaults.duration
  } = props;
  const slideCount = slides.length;
  const rootPrefixCls = (0, _configProvider.useCssScope)('carousel')('');
  const trackDescriptor = (0, _reactUtilities.useConstant)(() => transitionDescriptor);
  const manager = (0, _reactUtilities.useConstant)(() => new _CarouselManager.CarouselManager({
    trackDescriptor,
    defaultSlideIndex: (0, _clamp.default)(slideIndexProp !== null && slideIndexProp !== void 0 ? slideIndexProp : defaultSlideIndex, 0, slideCount - 1)
  }));
  manager.slideCount = slideCount;
  manager.duration = duration;
  manager.onSlideIndexChange = onSlideIndexChange;
  /**
   * States
   */
  const slide = (0, _shim.useSyncExternalStore)(manager.subscribe, React.useCallback(() => manager.state.slide, [manager]));
  React.useMemo(() => {
    // TODO 这里如果 onChange 的和 controlled value 不一致会 reset 回 controlled value，但是 controlled slide 引用会变，之后的逻辑不能依赖于 slide 的引用，而是需要做 deep equal
    if ((0, _util.isNonNullable)(slideIndexProp) && slideIndexProp !== slide.slideIndex) {
      manager.setSlideIndex(slideIndexProp, {
        triggerChange: false
      });
    }
  }, [slide.slideIndex, slideIndexProp, manager]);

  /**
   * ====================
   *   AutoPlay
   * ====================
   */
  const {
    play,
    pause,
    autoPlayPauseOnHover
  } = useCarouselAutoPlay({
    autoPlay
  }, {
    manager,
    slide
  });
  return {
    manager,
    contextValue: React.useMemo(() => ({
      rootPrefixCls,
      slide,
      slides,
      slideCount,
      duration,
      autoPlayPauseOnHover,
      play,
      pause
    }), [rootPrefixCls, slide, slides, slideCount, duration, autoPlayPauseOnHover, play, pause])
  };
};
exports.useCarousel = useCarousel;
/**
 * Carousel 组件全局状态
 */
const {
  Provider: CarouselContextProvider,
  useContext: useCarouselContext
} = (0, _reactUtilities.createContextHook)('CarouselContext', {
  strict: true
});
exports.useCarouselContext = useCarouselContext;
exports.CarouselContextProvider = CarouselContextProvider;
