// @universe-design/react/es/popover/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.u1guyajx{align-items:baseline;}
import u1gmpn6f from "./../.stylex/u1gmpn6f";
//.u4ybsrx{font-size:16px;}
import u1o367ev from "./../.stylex/u1o367ev";
//.uez10in{line-height:0;}
import u5p8ukz from "./../.stylex/u5p8ukz";
//.u7nlbp4{align-self:center;}
import u1mtnbey from "./../.stylex/u1mtnbey";
//.u19lj068{padding-right:8px;}
import u441rjq from "./../.stylex/u441rjq";
import React, { forwardRef, useMemo } from 'react';
import clsx from 'clsx';
import stylex from '@universe-design/stylex';
import { cssFlexBase, cssTypographyHeadline } from '../stylex';
import { useCssScope } from '../config-provider/hooks';
import InfoFilled from '@universe-design/icons-react/InfoFilled';
import SucceedFilled from '@universe-design/icons-react/SucceedFilled';
import WarningFilled from '@universe-design/icons-react/WarningFilled';
import ErrorFilled from '@universe-design/icons-react/ErrorFilled';
import { Tooltip } from '../tooltip/Tooltip';
import { getRenderPropValue } from '../util/getRenderPropValue';
import { FirstLineAlign } from '../shared/components/FirstLineAlign';
import './style';

// 正常、宽松、无任何内容（并且无填充）

/**
 * @public
 * @docgen
 */

const iconMap = {
  success: SucceedFilled,
  warning: WarningFilled,
  info: InfoFilled,
  error: ErrorFilled
};
export const Popover = /*#__PURE__*/forwardRef(function (props, ref) {
  const {
    mode = 'normal',
    prefixCls,
    title,
    content,
    overlayClassName,
    icon,
    mouseLeaveDelay = 100,
    ...restProps
  } = props;
  const cssScope = useCssScope('popover', prefixCls);
  const contentCls = cssScope('__overlay-content');
  const titleBeforeProps = useMemo(() => {
    if (!icon) {
      return;
    }
    const cssIconWrapper = {
      "props": {
        "z189sj": u441rjq
      }
    };
    const cssIcon = {
      "props": {
        "Be2twd7": u1o367ev,
        "Bg96gwp": u5p8ukz,
        "qb2dma": u1mtnbey
      }
    };
    if (typeof icon === 'string' && (icon === 'success' || icon === 'warning' || icon === 'info' || icon === 'error')) {
      const Icon = iconMap[icon];
      return {
        className: `${cssScope()}-icon-${icon}`,
        sx: [cssIconWrapper],
        children: /*#__PURE__*/React.createElement(Icon, {
          className: stylex(cssIcon)
        })
      };
    }
    return {
      className: `${cssScope()}-icon-customer`,
      sx: [cssIconWrapper],
      children: /*#__PURE__*/React.createElement("span", {
        className: clsx(`${cssScope()}-icon-customer-wrapper`, stylex(cssIcon))
      }, icon)
    };
  }, [cssScope, icon]);
  const popupNode = useMemo(() => {
    if (!title && !content) {
      return null;
    }

    // 纯净模式只渲染 content，并且移除了 padding 相关的 style
    if (mode === 'pure') {
      return content && /*#__PURE__*/React.createElement("div", {
        className: contentCls
      }, getRenderPropValue(content));
    }
    return /*#__PURE__*/React.createElement(React.Fragment, null, title && /*#__PURE__*/React.createElement(FirstLineAlign, {
      className: cssScope('__overlay-title'),
      sx: [cssFlexBase, cssTypographyHeadline, {
        "props": {
          "Bt984gj": u1gmpn6f
        }
      }],
      beforeProps: titleBeforeProps
    }, getRenderPropValue(title)), content && /*#__PURE__*/React.createElement("div", {
      className: contentCls
    }, getRenderPropValue(content)));
  }, [content, contentCls, cssScope, mode, title, titleBeforeProps]);
  return /*#__PURE__*/React.createElement(Tooltip, _extends({
    ref: ref,
    prefixCls: cssScope(),
    overlay: popupNode,
    overlayClassName: clsx(overlayClassName, cssScope(`--mode-${mode}`)),
    mouseLeaveDelay: mouseLeaveDelay,
    arrowHasStroke: true
  }, restProps));
});
if (process.env.NODE_ENV !== "production") {
  Popover.displayName = 'Popover';
}
