// @universe-design/react/es/tree-select/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
/* eslint-disable react-hooks/rules-of-hooks */
/* eslint-disable max-lines-per-function */

import React, { useCallback, useContext, useMemo } from 'react';
import { COMPONENT_VERSION_ID, useCssScope } from '../config-provider';
import { SelectActionType } from '../select/internal.interface';
import { SelectContext, selectGenerator } from '../select';
import { defaultFilterTreeNode } from './common';
import { ShowCheckedStrategy, traverseTreeConfig, TreeNode, getTreeNodeConfigPath } from '../tree';
import { genArray, isFunction, isNonNullable } from '../util';
import { treeChildren2TreeNodes } from '../tree/common';
import { genDoublyTreeNodeMap } from '../tree/util';
import { TreeCheckedProvider } from '../tree/hooks/TreeCheckedContext';
import { strategyMap } from '../tree/hooks/checkStrategy';
import { useTreeCheck } from '../tree/hooks/useTreeCheck';
import { useTree } from '../tree/hooks/useTree';
import { TreeContainer } from '../tree/TreeContainer';
import { TreeContextProvider } from '../tree/hooks/TreeContext';
import { SelectOverlayContainer } from '../select/SelectOverlayContainer';
import './style';
const RefSelect = selectGenerator({
  selectorCssScope: 'select',
  dropdownCssScope: 'select',
  defaultOptionLabelProp: option => {
    var _ref, _option$label;
    return (_ref = (_option$label = option.label) !== null && _option$label !== void 0 ? _option$label : option.title) !== null && _ref !== void 0 ? _ref : option.value;
  }
});

/**
 *   因为 tree 和 select-tree的value参数不一样， 所以需要将 tree-select的value 格式化为tree的value
 *   string[] -> string[] | {checked: string[], unchecked: string[]}
 */

function toCheckStrictlyValue(value) {
  if (Array.isArray(value)) {
    return {
      checked: value,
      halfChecked: []
    };
  }
  return {
    checked: [],
    halfChecked: []
  };
}
const TreeSelectOverlay = _ref2 => {
  var _treeData$length;
  let {
    multiple,
    treeCheckable,
    treeData,
    treeDefaultExpandAll,
    treeDefaultExpandedKeys,
    treeExpandedKeys,
    onTreeExpand,
    listHeight,
    filterTreeNode: filterTreeNodeProp,
    showCheckedStrategy,
    treeCheckStrictly,
    switcherIcon,
    loadData,
    virtual,
    nodeRenderer,
    treeRef,
    treeLoadedKeys,
    onTreeLoad
  } = _ref2;
  const cssScope = useCssScope('treeSelect');
  const {
    inputValue,
    valueSet,
    set
  } = useContext(SelectContext);
  let treeProps = {
    ref: treeRef,
    searchValue: inputValue,
    defaultExpandAll: treeDefaultExpandAll,
    expandedKeys: treeExpandedKeys,
    defaultExpandedKeys: treeDefaultExpandedKeys,
    onExpand: onTreeExpand,
    multiple,
    selectable: !treeCheckable,
    checkable: treeCheckable,
    showChecker: !treeCheckable,
    treeData,
    selectedKeys: treeCheckable ? [] : [...valueSet],
    nodeRenderer,
    onSelect: (keys, _action) => {
      const action = {
        type: _action.selected ? SelectActionType.Select : SelectActionType.DeSelect,
        payload: _action.key
      };
      set(keys, {
        action
      });
    },
    // ==================== Filter Tree Node ====================
    filterTreeNode: useCallback(node => {
      if (filterTreeNodeProp === false) {
        return true;
      } else if (isFunction(filterTreeNodeProp)) {
        return filterTreeNodeProp(inputValue, node);
      }
      return defaultFilterTreeNode(inputValue, node);
    }, [filterTreeNodeProp, inputValue]),
    showCheckedStrategy,
    switcherIcon,
    loadData,
    height: listHeight - 3 * 2 - 1 * 2,
    // 减去 padding border
    virtual,
    loadedKeys: treeLoadedKeys,
    onLoad: onTreeLoad
  };
  if (!treeCheckStrictly) {
    treeProps = {
      ...treeProps,
      checkedKeys: treeCheckable ? [...valueSet] : [],
      checkStrictly: false,
      onCheck: (keys, _action) => {
        const action = {
          type: _action.checked ? SelectActionType.Checked : SelectActionType.UnChecked,
          payload: _action.key
        };
        set(keys, {
          closeOverlay: false,
          action
        });
      }
    };
  } else {
    treeProps = {
      ...treeProps,
      defaultCheckedKeys: {
        checked: [],
        halfChecked: []
      },
      checkedKeys: treeCheckable ? {
        checked: [...valueSet],
        halfChecked: []
      } : {
        checked: [],
        halfChecked: []
      },
      checkStrictly: true,
      onCheck: (keys, _action) => {
        const action = {
          type: _action.checked ? SelectActionType.Checked : SelectActionType.UnChecked,
          payload: _action.key
        };
        set(keys.checked, {
          closeOverlay: false,
          action
        });
      }
    };
  }
  const treeCtxValue = useTree(treeProps, treeRef);
  const menuNode = /*#__PURE__*/React.createElement(TreeContextProvider, {
    value: treeCtxValue
  }, /*#__PURE__*/React.createElement(TreeContainer, null));
  return /*#__PURE__*/React.createElement(SelectOverlayContainer, {
    className: cssScope('__overlay'),
    optionsLength: (_treeData$length = treeData === null || treeData === void 0 ? void 0 : treeData.length) !== null && _treeData$length !== void 0 ? _treeData$length : 0,
    filteredOptionsLength: treeCtxValue.filteredTreeNodes.size,
    menu: menuNode
  });
};

/**
 * @docgen
 */

const _TreeSelect = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _props$multiple;
  const {
    treeData: treeDataProp,
    treeCheckable = false,
    treeDefaultExpandAll = false,
    treeDefaultExpandedKeys = [],
    treeExpandedKeys,
    onTreeExpand,
    listHeight = 232,
    filterTreeNode: filterTreeNodeProp = true,
    showCheckedStrategy = ShowCheckedStrategy.SHOW_CHILD,
    treeCheckStrictly = false,
    switcherIcon,
    loadData,
    children,
    virtual,
    nodeRenderer,
    treeLoadedKeys,
    onTreeLoad,
    ...baseSelectProps
  } = props;
  const selectRef = React.useRef(null);
  const treeRef = React.useRef(null);
  const treeData = useMemo(() => treeDataProp !== null && treeDataProp !== void 0 ? treeDataProp : treeChildren2TreeNodes(children), [children, treeDataProp]);
  const {
    doublyTreeNodeMap
  } = useMemo(() => genDoublyTreeNodeMap(treeData || []), [treeData]);
  const treeCheckCtx = useTreeCheck({
    doublyTreeNodeMap,
    checkStrategy: strategyMap[showCheckedStrategy],
    ...(treeCheckStrictly ? {
      checkStrictly: true,
      defaultCheckedKeys: toCheckStrictlyValue(props.defaultValue),
      ...('value' in props ? {
        checkedKeys: toCheckStrictlyValue(props.value)
      } : {})
    } : {
      checkStrictly: false,
      defaultCheckedKeys: genArray(props.defaultValue),
      ...('value' in props ? {
        checkedKeys: genArray(props.value)
      } : {})
    })
  }, {
    skipped: !treeCheckable
  });
  const {
    checkedSet,
    setCheckedKeys
  } = treeCheckCtx;
  const checkedKeys = React.useMemo(() => Object.keys(checkedSet).filter(key => checkedSet[key]), [checkedSet]);

  /**
   * Select Props
   */
  let multiple = (_props$multiple = props.multiple) !== null && _props$multiple !== void 0 ? _props$multiple : false;
  if (treeCheckable) {
    multiple = true;
  }
  baseSelectProps.multiple = multiple;
  const internalOptions = React.useMemo(() => {
    const options = [];
    traverseTreeConfig(treeData, node => {
      options.push(node);
    });
    return options;
  }, [treeData]);
  const overlayNode = /*#__PURE__*/React.createElement(TreeCheckedProvider, {
    value: treeCheckCtx
  }, /*#__PURE__*/React.createElement(TreeSelectOverlay, {
    treeRef: treeRef,
    treeData,
    treeCheckable,
    treeDefaultExpandAll,
    treeDefaultExpandedKeys,
    treeExpandedKeys,
    onTreeExpand,
    listHeight,
    multiple,
    filterTreeNode: filterTreeNodeProp,
    showCheckedStrategy,
    treeCheckStrictly,
    switcherIcon,
    loadData,
    virtual,
    nodeRenderer,
    treeLoadedKeys,
    onTreeLoad
  }));
  let selectProps;
  if (baseSelectProps.multiple) {
    if (treeCheckable) {
      if ('value' in props) {
        // 受控模式
        baseSelectProps.value = checkedKeys;
      } else {
        // 非受控模式
        baseSelectProps.defaultValue = checkedKeys;
      }
    }
    selectProps = {
      ...baseSelectProps,
      mode: 'multiple',
      options: internalOptions,
      overlayNode,
      onChange: (value, meta) => {
        if (treeCheckable) {
          const newValue = value;
          setCheckedKeys(treeCheckStrictly ? toCheckStrictlyValue(newValue) : genArray(newValue));
        }
        if (baseSelectProps.onChange) {
          const path = value.length ? value.reduce((a, b, index) => {
            const _path = getTreeNodeConfigPath(doublyTreeNodeMap[b]);
            a[index] = _path;
            return a;
          }, []) : [];
          baseSelectProps.onChange(value, {
            ...meta,
            path
          });
        }
      }
    };
  } else {
    selectProps = {
      ...baseSelectProps,
      mode: 'single',
      options: internalOptions,
      overlayNode,
      onChange: (value, meta) => {
        if (baseSelectProps.onChange) {
          const path = getTreeNodeConfigPath(isNonNullable(value) ? doublyTreeNodeMap[value] : null);
          baseSelectProps.onChange(value, {
            ...meta,
            path
          });
        }
      }
    };
  }
  React.useImperativeHandle(ref, () => ({
    get input() {
      var _selectRef$current;
      return (_selectRef$current = selectRef.current) === null || _selectRef$current === void 0 ? void 0 : _selectRef$current.input;
    },
    blur() {
      var _selectRef$current2;
      (_selectRef$current2 = selectRef.current) === null || _selectRef$current2 === void 0 ? void 0 : _selectRef$current2.blur();
    },
    focus() {
      var _selectRef$current3;
      (_selectRef$current3 = selectRef.current) === null || _selectRef$current3 === void 0 ? void 0 : _selectRef$current3.focus();
    },
    scrollIntoView() {
      var _treeRef$current;
      for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }
      (_treeRef$current = treeRef.current) === null || _treeRef$current === void 0 ? void 0 : _treeRef$current.scrollIntoView(...args);
    },
    toggleExpanded(key) {
      var _treeRef$current2;
      (_treeRef$current2 = treeRef.current) === null || _treeRef$current2 === void 0 ? void 0 : _treeRef$current2.toggleExpanded(key);
    }
  }));
  return /*#__PURE__*/React.createElement(RefSelect, _extends({
    ref: selectRef
  }, selectProps));
});
export const TreeSelect = Object.assign(_TreeSelect, {
  SHOW_ALL: ShowCheckedStrategy.SHOW_ALL,
  SHOW_PARENT: ShowCheckedStrategy.SHOW_PARENT,
  SHOW_CHILD: ShowCheckedStrategy.SHOW_CHILD,
  TreeNode,
  [COMPONENT_VERSION_ID]: "3.28.3",
  displayName: 'TreeSelect'
});
export * from './interface';
