// @universe-design/react/es/transfer/utils/index.js
import escapeRegExp from 'lodash/escapeRegExp';
export let TransferListType = /*#__PURE__*/function (TransferListType) {
  TransferListType[TransferListType["CAPION"] = 0] = "CAPION";
  TransferListType[TransferListType["OPTION"] = 1] = "OPTION";
  return TransferListType;
}({});
/**
 * List 只接受一维数组，这里是拍平后的数据结构
 */

/**
 * 将带分组的嵌套数据展平，
 * 因为 List 组件只接受一维数组
 */
export const flatTransferGroups = (options, groupWeakMap, showGroup) => {
  if (!showGroup) {
    return options.map(op => ({
      option: op,
      type: TransferListType.OPTION
    }));
  }
  const grouped = new Map();
  const groupSet = new Set();
  options.forEach(op => {
    const groupEntity = groupWeakMap.get(op);
    if (!groupEntity) {} else {
      groupSet.add(groupEntity);
      if (!grouped.has(groupEntity)) {
        grouped.set(groupEntity, []);
      }
      grouped.set(groupEntity, [...(grouped.get(groupEntity) || []), op]);
    }
  });
  const data = [];
  groupSet.forEach(groupEntity => {
    const groupTitle = groupEntity.title;
    data.push({
      group: {
        title: groupTitle,
        key: groupEntity.key
      },
      type: TransferListType.CAPION
    });
    data.push(...(grouped.get(groupEntity) || []).map(op => ({
      option: op,
      type: TransferListType.OPTION
    })));
  });
  return data;
};
export const defaultFilterOption = (searchValue, option) => {
  const keyword = typeof option.label === 'string' ? option.label : String(option.value);
  return !searchValue || new RegExp(escapeRegExp(searchValue), 'gi').test(keyword);
};
export const optionLabel = option => typeof option.label === 'string' ? option.label : String(option.value);
