// @universe-design/react/lib/date-picker/utils/ui.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.elementsContains = elementsContains;
exports.getInputSize = getInputSize;
exports.getStartOfPanel = void 0;
exports.getTimeProps = getTimeProps;
exports.isInHoverRange = isInHoverRange;
exports.isInRangeView = isInRangeView;
var _utils = require("../panels/base/utils");
function elementsContains(elements, target) {
  return elements.some(ele => ele && ele.contains(target));
}
function isInHoverRange(date, mode, hoverDate, rangeValue) {
  let index = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : 0;
  if (!hoverDate || !rangeValue) {
    return false;
  }
  const [startDate = null, endDate = null] = rangeValue;
  if (index === 0 && endDate) {
    return date.isSameOrAfter(hoverDate, mode) && date.isSameOrBefore(endDate, mode);
  }
  if (index === 1 && startDate) {
    return date.isSameOrAfter(startDate, mode) && date.isSameOrBefore(hoverDate, mode);
  }
  return false;
}
function getTimeProps(props) {
  const {
    format,
    showHour,
    showMinute,
    showSecond,
    use12Hours
  } = props;
  const showTimeObj = {
    ...props
  };
  if (format && typeof format === 'string') {
    if (!format.includes('s') && showSecond === undefined) {
      showTimeObj.showSecond = false;
    }
    if (!format.includes('m') && showMinute === undefined) {
      showTimeObj.showMinute = false;
    }
    if (!format.includes('H') && !format.includes('h') && showHour === undefined) {
      showTimeObj.showHour = false;
    }
    if ((format.includes('a') || format.includes('A')) && use12Hours === undefined) {
      showTimeObj.use12Hours = true;
    }
  }

  // if (picker === 'time') {
  //   return showTimeObj;
  // }

  // if (typeof format === 'function') {
  //   // format of showTime should use default when format is custom format function
  //   delete showTimeObj.format;
  // }

  return showTimeObj;
}
function getInputSize(format, generateConfig) {
  let textSize;
  if (typeof format === 'function') {
    textSize = format(generateConfig.getNow()).length;
  } else {
    textSize = format.length;
  }
  return Math.max(textSize + 2, 12);
}

/**
 * 获取单面板的开始时间
 */
const getStartOfPanel = (_value, mode) => {
  const value = _value.clone();
  if (mode === 'year') {
    return (0, _utils.getYearStart)(value).startOf('year');
  }
  return value.startOf(mode === 'month' ? 'year' : 'month');
};

/**
 * 判断是否在指定面板上
 */
exports.getStartOfPanel = getStartOfPanel;
function isInRangeView(
/**
 * 要判断的值
 */
_value,
/**
 * 面板值
 */
_pickerValues,
/**
 * 面板 mode
 */
mode) {
  let unit, startOfPanel, endOfPanel;
  const value = _value.clone();
  return _pickerValues.some(_pickerValue => {
    const pickerValue = _pickerValue.clone();
    if (mode === 'year') {
      startOfPanel = (0, _utils.getYearStart)(pickerValue).startOf('year');
      endOfPanel = startOfPanel.clone().add(_utils.YEAR_COUNT - 1, 'year').endOf('year');
    } else {
      switch (mode) {
        case 'month':
          unit = 'year';
          break;
        default:
          unit = 'month';
      }
      startOfPanel = pickerValue.clone().startOf(unit);
      endOfPanel = pickerValue.clone().endOf(unit);
    }
    return !value.isBefore(startOfPanel) && !value.isAfter(endOfPanel);
  });
}
