// @universe-design/react/es/space/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import clsx from 'clsx';
import React, { forwardRef } from 'react';
import { useCssScope } from '../config-provider/hooks';
import { toFlatArray } from '../util';
import './style';

/**
 * 间距尺寸
 * @public
 */

const spaceSize = {
  small: 8,
  medium: 16,
  large: 24
};
function getNumberSize(size) {
  return typeof size === 'string' ? spaceSize[size] : size || 0;
}
/**
 * @public
 */

export const Space = /*#__PURE__*/forwardRef(function (_ref, ref) {
  let {
    children,
    align,
    size = 'small',
    direction = 'horizontal',
    className,
    wrap = false,
    split,
    style,
    keepEmpty = false,
    inline = true,
    shouldItemConfinedToContainer = false,
    itemKey,
    shouldPlaceItemInFlowRoot = true,
    ...restProps
  } = _ref;
  const cssScope = useCssScope('space');
  const itemCssScope = useCssScope('space__item');
  const childNodes = toFlatArray(children, {
    keepEmpty
  });
  const [horizontalSize, verticalSize] = React.useMemo(() => (Array.isArray(size) ? size : [size, size]).map(getNumberSize), [size]);
  const latestIndex = childNodes.length - 1;
  if (childNodes.length === 0) {
    return null;
  }
  const mergedAlign = (() => {
    if (align) {
      return align;
    }
    if (direction === 'horizontal') {
      return 'center';
    }
    if (!shouldPlaceItemInFlowRoot) {
      return 'start';
    }
  })();
  return /*#__PURE__*/React.createElement("div", _extends({}, restProps, {
    ref: ref,
    className: clsx(cssScope(), cssScope(`-${direction}`), inline && cssScope('-inline'), mergedAlign && cssScope(`-align-${mergedAlign}`), className),
    style: {
      ...(wrap && {
        flexWrap: 'wrap'
      }),
      columnGap: horizontalSize / (split ? 2 : 1),
      rowGap: verticalSize,
      ...style
    }
  }), childNodes.map((node, index) => /*#__PURE__*/React.createElement(React.Fragment, {
    key: itemKey ? itemKey(index) : index
  }, shouldPlaceItemInFlowRoot ? /*#__PURE__*/React.createElement("div", {
    className: clsx(itemCssScope(), shouldItemConfinedToContainer && itemCssScope('-confined'))
  }, node) : node, index < latestIndex && split && /*#__PURE__*/React.createElement("span", {
    className: itemCssScope('__split')
  }, split))));
});
