// @universe-design/react/es/date-picker/panels/picker/Footer.js
import React from 'react';
import moment from 'moment';
import { useDatePickerContext } from '../../Context';
import { useCssScope, useLocale } from '../../../config-provider/hooks';
import { Button } from '../../../button';
import clsx from 'clsx';
export const DatePickerFooter = props => {
  const {
    global: globalLocale,
    components: {
      datePicker: locale
    }
  } = useLocale();
  const pickerFooterCssScope = useCssScope('picker__footer');
  const datePickerCtx = useDatePickerContext();
  const {
    renderExtraFooter,
    setValue,
    showToday,
    needConfirm,
    showNow
  } = datePickerCtx;
  if (!showToday && !showNow && !needConfirm && !renderExtraFooter) {
    return null;
  }
  return /*#__PURE__*/React.createElement("div", {
    className: clsx(pickerFooterCssScope(), pickerFooterCssScope('-container'))
  }, function () {
    var _props$disabledDate;
    if (renderExtraFooter) {
      return /*#__PURE__*/React.createElement("div", {
        key: "extra",
        className: pickerFooterCssScope('__extra')
      }, renderExtraFooter(datePickerCtx));
    }
    return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("div", {
      key: "default",
      className: clsx([pickerFooterCssScope('__date'), needConfirm && pickerFooterCssScope('__date--left')])
    }, showToday && /*#__PURE__*/React.createElement(Button, {
      type: "text",
      color: "primary",
      disabled: (_props$disabledDate = props.disabledDate) === null || _props$disabledDate === void 0 ? void 0 : _props$disabledDate.call(props, moment()),
      onClick: () => {
        setValue(moment(), !needConfirm);
      }
    }, locale.today), showNow && /*#__PURE__*/React.createElement(Button, {
      type: "text",
      color: "primary",
      onClick: () => {
        setValue(moment(), false);
      }
    }, locale.currentTime)), needConfirm && /*#__PURE__*/React.createElement("div", {
      key: "confirm",
      className: pickerFooterCssScope('-confirm-buttons')
    }, /*#__PURE__*/React.createElement(Button, {
      size: "small",
      onClick: props.onCancel
    }, globalLocale.cancelText), /*#__PURE__*/React.createElement(Button, {
      size: "small",
      type: "primary",
      disabled: !props.value,
      onClick: props.onConfirm
    }, globalLocale.okText)));
  }());
};
