// @universe-design/react/es/date-picker/panels/picker/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import clsx from 'clsx';
import moment from 'moment';
import React, { useCallback } from 'react';
import noop from 'lodash/noop';
import { useCssScope, useLocale } from '../../../config-provider/hooks';
import TimePickerPanel from '../../../shared/components/picker/Comp/panels/time-panel';
import generateConfig from '../../../shared/components/picker/generate/moment';
import { DatePickerContextProvider, useDatePickerContext } from '../../Context';
import { DatePickerFooter } from './Footer';
import { getTimeProps } from '../../utils/ui';
import { mergeDateAndTime } from '../../utils/date';
import { PickerContext } from '../../../shared/components/picker/Context';
import useSyncState from '../../../hooks/useSyncState';
import { DateCalendar } from '../DateCalendar/DateCalendar';
import '../../style';

/**
 * @privateRemarks rangedValue不会对视图定位产生影响，value会
 * @public
 * @docgen
 * 日期选择面板的属性
 */

export const isDateValueEqual = (a, b) => {
  if (a === b || a !== null && a !== void 0 && a.isSame(b)) {
    return true;
  }
  return false;
};
/**
 * @deprecated
 * 推荐使用 `DateCalendar` 或 `DateCalendarUI` 组件
 */
export function PickerPanel(props) {
  const {
    className,
    style,
    value,
    mode = 'date',
    footer = true,
    defaultViewDate,
    format,
    disabledDate,
    // footerProps
    onCancel,
    onConfirm,
    // bodyProps
    onSelect,
    onDateMouseEnter: onDateMouseEnterProp,
    onViewDateChange: onPickerValueChange,
    dateRender,
    ...restBodyProps
  } = props;
  const prefixCls = useCssScope('picker')();
  const datePanelPrefixCls = `${prefixCls}-date-panel`;
  const datePickerContext = useDatePickerContext();
  const {
    setValue,
    showTime,
    mode: localMode,
    setMode: setLocalMode
  } = datePickerContext;
  const timeOperationRef = React.useRef({});
  const timePanelFormat = typeof format === 'function' ? undefined : format;
  const timePickerProps = showTime ? getTimeProps({
    ...(typeof showTime === 'object' ? showTime : {}),
    format: timePanelFormat
  }) : {};
  const defaultTimeValue = timePickerProps.defaultValue || moment();

  // 当前面板切换到哪个时间点，并不是已选中的值或未提交的值
  const [pickerValue = moment(), setViewDateState] = useSyncState(value !== null && value !== void 0 ? value : defaultViewDate, {
    isEqual: isDateValueEqual
  });
  const setViewDate = date => {
    onPickerValueChange === null || onPickerValueChange === void 0 ? void 0 : onPickerValueChange(date, {
      mode: localMode
    });
    setViewDateState(date);
  };
  const onDateCalendarSelect = date => {
    if (!date) {
      // 不存在此情况
      return;
    }
    const newDate = mergeDateAndTime(date, showTime ? value !== null && value !== void 0 ? value : defaultTimeValue : null);
    onSelect === null || onSelect === void 0 ? void 0 : onSelect(newDate);
    if (mode !== localMode) {
      setLocalMode(mode);
    }
  };
  const onSelectTime = useCallback(time => {
    onSelect === null || onSelect === void 0 ? void 0 : onSelect(time);
  }, [onSelect]);
  const locale = useLocale();
  const onDateMouseEnter = useCallback(v => {
    onDateMouseEnterProp === null || onDateMouseEnterProp === void 0 ? void 0 : onDateMouseEnterProp(mergeDateAndTime(v, showTime ? value !== null && value !== void 0 ? value : defaultTimeValue : null));
  }, [defaultTimeValue, onDateMouseEnterProp, showTime, value]);
  return /*#__PURE__*/React.createElement("div", {
    className: className,
    style: style
  }, /*#__PURE__*/React.createElement("div", {
    className: clsx(`${datePanelPrefixCls}__wrapper`, showTime && `${datePanelPrefixCls}--withTime`)
  }, /*#__PURE__*/React.createElement(DateCalendar, _extends({}, restBodyProps, {
    onPickerValueChange: setViewDate,
    onDateMouseEnter: onDateMouseEnter,
    value: value || undefined,
    mode: localMode,
    pickerValue: pickerValue,
    onSelect: d => onDateCalendarSelect === null || onDateCalendarSelect === void 0 ? void 0 : onDateCalendarSelect(d),
    disabledDate: disabledDate
  }, dateRender ? {
    cellRenderer: (date, _ref) => {
      let {
        viewMode,
        defaultNode
      } = _ref;
      return dateRender(date, viewMode, defaultNode);
    }
  } : {})), showTime && /*#__PURE__*/React.createElement(PickerContext.Provider, {
    value: {
      value: value !== null && value !== void 0 ? value : undefined,
      mode: 'time',
      setMode: noop,
      setValue,
      setOpened: noop,
      inDatePicker: true
    }
  }, /*#__PURE__*/React.createElement(TimePickerPanel, _extends({
    locale: locale.components.picker
  }, timePickerProps, {
    picker: "time",
    mode: "time",
    prefixCls: prefixCls,
    generateConfig: generateConfig,
    onPanelChange: args => onSelectTime(args),
    operationRef: timeOperationRef
  })))), footer && /*#__PURE__*/React.createElement(DatePickerContextProvider, {
    value: {
      ...datePickerContext,
      setValue: (date, autoCommit) => {
        setViewDate(date);
        setValue(date, autoCommit);
      }
    }
  }, /*#__PURE__*/React.createElement(DatePickerFooter, {
    value: value,
    onCancel: onCancel,
    onConfirm: onConfirm,
    disabledDate: disabledDate
  })));
}
export default PickerPanel;
