// @universe-design/react/src/shared/components/popper-sidecar/FloatingUIPopperSidecar.tsx
import * as React from 'react';
import type { PopperProps } from '../popper';
import { PopperArrow, Popper, ZERO_TIMEOUT } from '../popper';
import { deprecatedWarning, getFloatingUIPlacement, isNonNullable, warning } from '../../../util';
import type { PopperRef } from '../popper/PopperProvider';
import type { Placement } from '../popper/types';
import type { PopperSidecarProps } from './shared';
import { POPUP_BORDER_RADIUS, POPUP_ARROW_SIZE } from './shared';
import { unstable_useConfigReducedMotion as useConfigReducedMotion } from '../../../config-provider/ConfigReducedMotion';

export const FloatingUIPopperSidecar = React.forwardRef(function FloatingUIPopperSidecar(
  {
    // config props
    action,
    popupPlacement,
    showArrow,
    arrowHasStroke,
    popupMotion,
    defaultMainAxisOffsetWithoutArrow,
    // user props
    prefixCls,
    enableAnimation: enableAnimationProp,
    popupAlign: popupAlignProp,
    offset: offsetProp,
    flip: flipProp,
    shift: shiftProp,
    ...restProps
  }: PopperSidecarProps,
  ref: React.ForwardedRef<PopperRef>,
) {
  /**
   * Deprecated Warning
   */
  warning(!popupAlignProp, deprecatedWarning('FloatingUIPopperSidecar', 'popupAlign'));
  warning(!popupAlignProp?.offset, deprecatedWarning('FloatingUIPopperSidecar', 'popupAlign.offset', 'offset'));
  warning(
    popupAlignProp?.overflow?.adjustX === undefined,
    deprecatedWarning('FloatingUIPopperSidecar', 'popupAlign.overflow.adjustX', ['flip', 'shift']),
  );
  warning(
    popupAlignProp?.overflow?.adjustY === undefined,
    deprecatedWarning('FloatingUIPopperSidecar', 'popupAlign.overflow.adjustY', ['flip', 'shift']),
  );

  const popupAlign = popupAlignProp ?? {};
  const alignOffset = popupAlign?.offset;
  const points = popupAlign?.points;
  const adjustX = popupAlign?.overflow?.adjustX;
  const adjustY = popupAlign?.overflow?.adjustY;
  const alwaysByViewport = popupAlign?.overflow?.alwaysByViewport;

  const placement: Placement = (() => {
    if (isNonNullable(points)) {
      const [popupPoint, referencePoint] = points;
      // left
      if (popupPoint === 'tr' && referencePoint === 'tl') {
        return 'left-start';
      }
      if (popupPoint === 'cr' && referencePoint === 'cl') {
        return 'left';
      }
      if (popupPoint === 'br' && referencePoint === 'bl') {
        return 'left-end';
      }
      // right
      if (popupPoint === 'tl' && referencePoint === 'tr') {
        return 'right-start';
      }
      if (popupPoint === 'cl' && referencePoint === 'cr') {
        return 'right';
      }
      if (popupPoint === 'bl' && referencePoint === 'br') {
        return 'right-end';
      }
      // top
      if (popupPoint === 'bl' && referencePoint === 'tl') {
        return 'top-start';
      }
      if (popupPoint === 'bc' && referencePoint === 'tc') {
        return 'top';
      }
      if (popupPoint === 'br' && referencePoint === 'tr') {
        return 'top-end';
      }
      // bottom
      if (popupPoint === 'tl' && referencePoint === 'bl') {
        return 'bottom-start';
      }
      if (popupPoint === 'tc' && referencePoint === 'bc') {
        return 'bottom';
      }
      if (popupPoint === 'tr' && referencePoint === 'br') {
        return 'bottom-end';
      }
    }
    return getFloatingUIPlacement(popupPlacement);
  })();

  const isTop = placement.startsWith('top');
  const isBottom = placement.startsWith('bottom');
  const isVertical = isTop || isBottom;

  const flip: PopperProps['flip'] = React.useMemo(() => {
    if (isNonNullable(flipProp)) {
      return flipProp;
    }
    if (isNonNullable(adjustX) || isNonNullable(adjustY)) {
      const flipAxisOptions = isVertical
        ? {
            // https://floating-ui.com/docs/flip#mainaxis
            mainAxis: adjustY !== undefined ? Boolean(adjustY) : true, // 如果 vertical，主轴是 y，对应 adjustY
            crossAxis: adjustX !== undefined ? Boolean(adjustX) : true, // crossAxis flip 代表的是 xx-start -> xx-end 或反之
          }
        : {
            mainAxis: adjustX !== undefined ? Boolean(adjustX) : true,
            crossAxis: adjustY !== undefined ? Boolean(adjustY) : true,
          };
      return flipAxisOptions;
    }
  }, [adjustX, adjustY, flipProp, isVertical]);

  const shift: PopperProps['shift'] = React.useMemo(() => {
    if (isNonNullable(shiftProp)) {
      return shiftProp;
    }
    if (isNonNullable(adjustX) || isNonNullable(adjustY)) {
      const shiftAxisOptions = isVertical
        ? {
            // https://floating-ui.com/docs/shift#mainaxis
            mainAxis: adjustX !== undefined ? Boolean(adjustX) : true, // 如果 vertical，主轴是 x，我们不开启 crossAxis 的 shift（因为会盖住 reference），这种情况我们通过 flip + 内部滚动来解决
          }
        : {
            mainAxis: adjustY !== undefined ? Boolean(adjustY) : true,
          };
      return shiftAxisOptions;
    }
  }, [adjustX, adjustY, isVertical, shiftProp]);

  const offset: PopperProps['offset'] = React.useMemo(() => {
    if (isNonNullable(alignOffset)) {
      const [x, y] = alignOffset;
      if (placement.startsWith('right')) {
        return { mainAxis: x, crossAxis: y };
      }
      if (placement.startsWith('left')) {
        return { mainAxis: -x, crossAxis: y };
      }
      if (placement.startsWith('bottom')) {
        return { mainAxis: y, crossAxis: x };
      }
      if (placement.startsWith('top')) {
        return { mainAxis: -y, crossAxis: x };
      }
    }
    if (isNonNullable(offsetProp)) {
      return offsetProp;
    }
    return (showArrow ? POPUP_ARROW_SIZE.y : 0) + defaultMainAxisOffsetWithoutArrow;
  }, [alignOffset, defaultMainAxisOffsetWithoutArrow, offsetProp, placement, showArrow]);

  const enableAnimation = !useConfigReducedMotion(
    isNonNullable(enableAnimationProp) ? !enableAnimationProp : undefined,
  );

  return (
    <Popper
      ref={ref}
      {...{
        prefixCls,
        placement,
        trigger: action,
        arrow: showArrow ? <PopperArrow hasStroke={arrowHasStroke} /> : null,
        timeout: enableAnimation ? undefined : ZERO_TIMEOUT,
        offset,
        // 没有 arrow 的时候让 arrowPadding 是 0，这样可以跳过 arrow middleware，respect 安全边距而非箭头居中 https://github.com/floating-ui/floating-ui/pull/1496
        arrowPadding: showArrow ? POPUP_BORDER_RADIUS + POPUP_ARROW_SIZE.x : 0,
        flip,
        shift,
        rootBoundary: alwaysByViewport === false ? 'document' : undefined,
      }}
      {...restProps}
    />
  );
});

if (__DEV__) {
  FloatingUIPopperSidecar.displayName = 'FloatingUIPopperSidecar';
}
