// @universe-design/react/src/shared/components/popper-sidecar/RcTriggerPopperSidecar.tsx
import * as React from 'react';
import type { PopperRef } from '../popper/PopperProvider';
import type { BuildInPlacements, UDAlignType, UdTRiggerArrow } from '../trigger';
import { UDTrigger } from '../trigger';
import { StreamlineArrow } from '../trigger/StreamlineArrow';
import type { PopperSidecarProps } from './shared';
import { POPUP_ARROW_SIZE, POPUP_BORDER_RADIUS } from './shared';
import { devWarning, isNonNullable, warning } from '../../../util';
import { unstable_useConfigReducedMotion as useConfigReducedMotion } from '../../../config-provider/ConfigReducedMotion';
import type { PopperProps } from '../..';

export const getPlacements = ({
  offset: offsetProp,
  defaultOffset,
}: {
  offset: PopperProps['offset'];
  defaultOffset: number;
}): BuildInPlacements => {
  let offset = offsetProp;
  if (offset === undefined || typeof offset !== 'number') {
    offset = defaultOffset;
  }

  const defaultAlign: UDAlignType = {
    targetOffset: [0, 0],
    ignoreShake: true,
  };

  const overflow: UDAlignType['overflow'] = {
    adjustX: true,
    adjustY: true,
    alwaysByViewport: false,
  };

  const left = {
    points: ['cr', 'cl'],
    overflow,
    offset: [-offset, 0],
    ...defaultAlign,
  };
  const right = {
    points: ['cl', 'cr'],
    overflow,
    offset: [offset, 0],
    ...defaultAlign,
  };
  const top = {
    points: ['bc', 'tc'],
    overflow,
    offset: [0, -offset],
    ...defaultAlign,
  };
  const bottom = {
    points: ['tc', 'bc'],
    overflow,
    offset: [0, offset],
    ...defaultAlign,
  };
  const topLeft = {
    points: ['bl', 'tl'],
    overflow,
    offset: [0, -offset],
    ...defaultAlign,
  };
  const leftTop = {
    points: ['tr', 'tl'],
    overflow,
    offset: [-offset, 0],
    ...defaultAlign,
  };
  const topRight = {
    points: ['br', 'tr'],
    overflow,
    offset: [0, -offset],
    ...defaultAlign,
  };
  const rightTop = {
    points: ['tl', 'tr'],
    overflow,
    offset: [offset, 0],
    ...defaultAlign,
  };
  const bottomRight = {
    points: ['tr', 'br'],
    overflow,
    offset: [0, offset],
    ...defaultAlign,
  };
  const rightBottom = {
    points: ['bl', 'br'],
    overflow,
    offset: [offset, 0],
    ...defaultAlign,
  };
  const bottomLeft = {
    points: ['tl', 'bl'],
    overflow,
    offset: [0, offset],
    ...defaultAlign,
  };
  const leftBottom = {
    points: ['br', 'bl'],
    overflow,
    offset: [-offset, 0],
    ...defaultAlign,
  };

  return {
    // legacy
    leftCenter: left,
    rightCenter: right,
    topCenter: top,
    bottomCenter: bottom,
    // in use
    left,
    right,
    top,
    bottom,
    topLeft,
    leftTop,
    topRight,
    rightTop,
    bottomRight,
    rightBottom,
    bottomLeft,
    leftBottom,
  };
};

export const RcTriggerPopperSidecar = React.forwardRef(function RcTriggerPopperSidecar(
  {
    // config props
    action,
    popupPlacement,
    showArrow,
    arrowHasStroke,
    popupMotion,
    defaultMainAxisOffsetWithoutArrow,
    // user props
    prefixCls,
    enableAnimation: enableAnimationProp,
    offset,
    // omit new popper props
    flip: flipProp,
    shift: shiftProp,
    ...restProps
  }: PopperSidecarProps,
  ref: React.ForwardedRef<PopperRef>,
) {
  /**
   * Deprecated Warning
   */
  warning(
    false,
    devWarning(
      'RcTriggerPopperSidecar',
      // eslint-disable-next-line max-len
      'Detected that `RcTriggerPopperSidecar` is being used. `RcTriggerPopperSidecar` does not support follow scrolling, safe margin and other capabilities, and will be removed in the next major version, It is recommended to use `FloatingUIPopperSidecar` instead partialy. The usage is as follows `<ConfigProvider PopperSidecar={FloatingUIPopperSidecar} />`. Learn more about what we do here: https://bytedance.feishu.cn/wiki/wikcnG3DTElZpfwQqB88ArZQiKe',
    ),
  );

  const builtinPlacements = React.useMemo(
    () =>
      getPlacements({
        offset,
        defaultOffset: (showArrow ? POPUP_ARROW_SIZE.y : 0) + defaultMainAxisOffsetWithoutArrow,
      }),
    [defaultMainAxisOffsetWithoutArrow, offset, showArrow],
  );

  const arrow: UdTRiggerArrow | undefined = React.useMemo(
    () =>
      showArrow
        ? {
            size: POPUP_ARROW_SIZE.x,
            safeMarginX: POPUP_BORDER_RADIUS,
            safeMarginY: POPUP_BORDER_RADIUS,
            content: ({ placement: arrowPlacement }) => {
              const arrowStyle = (() => {
                if (arrowPlacement === 'top') {
                  return {
                    transform: 'rotate(180deg)',
                    left: 0,
                    bottom: 0,
                  };
                }
                if (arrowPlacement === 'bottom') {
                  return {
                    left: 0,
                    top: 0,
                  };
                }
                if (arrowPlacement === 'left') {
                  return {
                    transform: 'rotate(180deg)',
                    right: 0,
                  };
                }
                if (arrowPlacement === 'right') {
                  return {
                    left: 0,
                  };
                }
              })();
              return (
                <StreamlineArrow
                  className={`${prefixCls}__arrow__streamline`}
                  style={{
                    position: 'absolute',
                    ...arrowStyle,
                  }}
                  direction={['left', 'right'].includes(arrowPlacement) ? 'vertical' : 'horizontal'}
                  hasStroke={arrowHasStroke}
                />
              );
            },
          }
        : undefined,
    [arrowHasStroke, prefixCls, showArrow],
  );

  const enableAnimation = !useConfigReducedMotion(
    isNonNullable(enableAnimationProp) ? !enableAnimationProp : undefined,
  );

  return (
    <UDTrigger
      ref={ref as any}
      {...{
        prefixCls,
        popupPlacement,
        builtinPlacements,
        action,
        arrow,
        // 这里必须要提供一个 duration 0s 的 animation，因为 rc-motion 会根据 animationEnd 去判断 open 过程的结束，没有对应的 animation 会导致 popup 一直处于 pointer-events: none 导致用户无法移入 popup
        // https://code.byted.org/universe-ui/react/issues/982
        popupMotion: enableAnimation ? popupMotion : 'noop',
      }}
      {...restProps}
    />
  );
});

if (__DEV__) {
  RcTriggerPopperSidecar.displayName = 'RcTriggerPopperSidecar';
}
