// @universe-design/react-aurora/lib-commonjs/aurora-background/AuroraBlock.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AuroraBlock = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _immer = _interopRequireDefault(require("immer"));
var _react = _interopRequireDefault(require("react"));
var _Ellipse = require("./Ellipse");
var _defaults = require("./defaults");
var _clsx = _interopRequireDefault(require("clsx"));
require("./styles/gradient-block.less");
/**
 * @public
 *  极光背景组件的属性
 */

/**
 * @public
 */
const AuroraBlock = exports.AuroraBlock = /*#__PURE__*/_react.default.forwardRef((props, ref) => {
  const {
    mainEllipse: mainEllipseProp,
    subEllipse: subEllipseProp,
    reflectEllipse: reflectEllipseProp,
    speed = 1,
    noAnimation = false,
    className,
    ...rest
  } = props;
  const ellipseConfig = _react.default.useMemo(() => ({
    ..._defaults.Default_Aurora_Config,
    mainEllipse: {
      ..._defaults.Default_Aurora_Config.mainEllipse,
      ...mainEllipseProp
    },
    subEllipse: {
      ..._defaults.Default_Aurora_Config.subEllipse,
      ...subEllipseProp
    },
    reflectEllipse: {
      ..._defaults.Default_Aurora_Config.reflectEllipse,
      ...reflectEllipseProp
    }
  }), [mainEllipseProp, reflectEllipseProp, subEllipseProp]);
  const [ellipse, setEllipse] = _react.default.useState(ellipseConfig);
  const cancelId = _react.default.useRef();
  const firstUpdateTimeRef = _react.default.useRef();
  const updatePosition = (_ellipse, zoomRatio) => ({
    ..._ellipse,
    rX: _ellipse.rX + _ellipse.deltaRX * zoomRatio,
    rY: _ellipse.rY + _ellipse.deltaRY * zoomRatio
  });
  _react.default.useLayoutEffect(() => {
    if (noAnimation) {
      return;
    }
    let zoomRatio = 0;
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    cancelId.current = window.setInterval(() => {
      const now = Date.now();
      if (firstUpdateTimeRef.current === undefined) {
        firstUpdateTimeRef.current = now;
      } else {
        const elapsed = now - firstUpdateTimeRef.current;
        zoomRatio = Math.sin(elapsed / _defaults.Default_Animation_Time * 2 * Math.PI * speed);
      }
      setEllipse((0, _immer.default)(draft => {
        draft.mainEllipse = updatePosition(ellipseConfig.mainEllipse, zoomRatio);
        draft.subEllipse = updatePosition(ellipseConfig.subEllipse, zoomRatio);
        draft.reflectEllipse = updatePosition(ellipseConfig.reflectEllipse, zoomRatio);
      }));
    }, 16);
    return () => {
      cancelId.current && window.clearInterval(cancelId.current);
    };
  }, [speed, ellipseConfig.mainEllipse, ellipseConfig.subEllipse, ellipseConfig.reflectEllipse, noAnimation]);
  return /*#__PURE__*/_react.default.createElement("svg", (0, _extends2.default)({
    className: (0, _clsx.default)(className, 'udx__aurora-block'),
    viewBox: "0 0 375 250",
    ref: ref
  }, rest), /*#__PURE__*/_react.default.createElement(_Ellipse.Ellipse, (0, _extends2.default)({}, ellipse.mainEllipse, {
    filter: "url(#udx_aurora_main)"
  })), /*#__PURE__*/_react.default.createElement(_Ellipse.Ellipse, (0, _extends2.default)({}, ellipse.subEllipse, {
    filter: "url(#udx_aurora_sub)"
  })), /*#__PURE__*/_react.default.createElement(_Ellipse.Ellipse, (0, _extends2.default)({}, ellipse.reflectEllipse, {
    filter: "url(#udx_aurora_reflection)"
  })), /*#__PURE__*/_react.default.createElement("defs", null, /*#__PURE__*/_react.default.createElement("filter", {
    id: "udx_aurora_main",
    filterUnits: "userSpaceOnUse",
    "color-interpolation-filters": "sRGB"
  }, /*#__PURE__*/_react.default.createElement("feFlood", {
    "flood-opacity": "0",
    result: "BackgroundImageFix"
  }), /*#__PURE__*/_react.default.createElement("feBlend", {
    mode: "normal",
    in: "SourceGraphic",
    in2: "BackgroundImageFix",
    result: "shape"
  }), /*#__PURE__*/_react.default.createElement("feGaussianBlur", {
    stdDeviation: "40",
    result: "effect1_foregroundBlur_210_68150"
  })), /*#__PURE__*/_react.default.createElement("filter", {
    id: "udx_aurora_sub",
    filterUnits: "userSpaceOnUse",
    "color-interpolation-filters": "sRGB"
  }, /*#__PURE__*/_react.default.createElement("feFlood", {
    "flood-opacity": "0",
    result: "BackgroundImageFix"
  }), /*#__PURE__*/_react.default.createElement("feBlend", {
    mode: "normal",
    in: "SourceGraphic",
    in2: "BackgroundImageFix",
    result: "shape"
  }), /*#__PURE__*/_react.default.createElement("feGaussianBlur", {
    stdDeviation: "50",
    result: "effect1_foregroundBlur_210_68150"
  })), /*#__PURE__*/_react.default.createElement("filter", {
    id: "udx_aurora_reflection",
    filterUnits: "userSpaceOnUse",
    "color-interpolation-filters": "sRGB"
  }, /*#__PURE__*/_react.default.createElement("feFlood", {
    "flood-opacity": "0",
    result: "BackgroundImageFix"
  }), /*#__PURE__*/_react.default.createElement("feBlend", {
    mode: "normal",
    in: "SourceGraphic",
    in2: "BackgroundImageFix",
    result: "shape"
  }), /*#__PURE__*/_react.default.createElement("feGaussianBlur", {
    stdDeviation: "40",
    result: "effect1_foregroundBlur_210_68150"
  }))));
});
