// @universe-design/react/src/metrics/stories/homeless.stories.tsx
import * as React from 'react';
import { Button, Card, Col, Input, Metrics, Row } from '@universe-design/react';
import { MetricsGroup } from '../MetricsGroup';
import { PerformanceContent } from '../../_stories/common/PerformanceContent';

export default {
  title: 'Components/Metrics/临时示例', // 注意：这里需要修改成自己对应 Component name。
  component: Metrics,
};

export const Grid = () => (
  <MetricsGroup>
    <Row gutter={[12, 12]}>
      <Col xs={12} sm={12}>
        <Metrics
          mainTitle="高绩效流失率（180 天）"
          mainInfoIcon={{
            tooltip: {
              title: '默认启用字号缩小',
            },
          }}
          mainValue="25.4%"
        />
      </Col>
      <Col xs={12} sm={12}>
        <Metrics
          mainTitle="高绩效流失率（本绩效周期）"
          mainInfoIcon={{
            tooltip: {
              title: '默认启用字号缩小',
            },
          }}
          mainValue="15.9%"
        />
      </Col>
      <Col xs={12} sm={12}>
        <Metrics
          mainTitle="低绩效流失率（180 天）"
          mainInfoIcon={{
            tooltip: {
              title: '默认启用字号缩小',
            },
          }}
          mainValue="175.9%"
        />
      </Col>
      <Col xs={12} sm={12}>
        <Metrics
          mainTitle="低绩效流失率（本绩效周期）"
          mainInfoIcon={{
            tooltip: {
              title: '默认启用字号缩小',
            },
          }}
          mainValue="100.0%"
          secondary={{
            data: [
              {
                type: 'none',
                label: '本次 M-，上次 ≤ M-',
                value: '',
                color: 'danger',
              },
              {
                type: 'none',
                label: '一级部门目标：≥55%',
                value: '',
                color: 'success',
              },
            ],
          }}
        />
      </Col>
    </Row>
  </MetricsGroup>
);

export const Multiple = () => {
  type Item = {
    id: number;
    title: string;
    value: string;
    trend: number;
  };
  const [data, setData] = React.useState<Item[]>([
    { id: 1, title: '销售额', value: '¥1000000', trend: 15 },
    { id: 2, title: '订单量', value: '¥2000000', trend: -5 },
    { id: 3, title: '用户数', value: '¥3000000', trend: 20 },
    { id: 4, title: '转化率', value: '¥4000000', trend: 8 },
    {
      id: 5,
      title: '超长标题标题标题标题标题标题标题标题标题标题标题标题标题标题标题标题标题标题标题标题',
      value: '¥5000000',
      trend: -12,
    },
  ]);
  const [editingItem, setEditingItem] = React.useState<Item | null>(null);
  const [editValue, setEditValue] = React.useState('');

  const handleAdd = () => {
    const newItem = {
      id: Date.now(),
      title: '新指标',
      value: '¥0',
      trend: 0,
    };
    setData([...data, newItem]);
  };

  const handleDelete = (id: number) => {
    setData(data.filter(item => item.id !== id));
  };

  const handleEdit = (item: Item | null) => {
    setEditingItem(item);
    setEditValue(item!.title); // 修改为编辑 title
  };

  const handleSave = () => {
    if (!editingItem) {
      return;
    }
    setData(
      data.map(item =>
        item.id === editingItem.id
          ? { ...item, title: editValue } // 修改为保存 title
          : item,
      ),
    );
    setEditingItem(null);
  };

  return (
    <div>
      <Button onClick={handleAdd}>添加指标</Button>
      <div style={{ width: '900px' }}>
        <MetricsGroup>
          {data.map(item => (
            <Card key={item.id} style={{ width: '300px', display: 'inline-block', verticalAlign: 'top' }}>
              <Metrics
                mainTitle={
                  item.id === editingItem?.id ? (
                    <Input
                      value={editValue}
                      onChange={e => setEditValue(e.target.value)}
                      onBlur={handleSave}
                      onPressEnter={handleSave}
                      autoFocus
                    />
                  ) : (
                    item.title
                  )
                }
                mainValue={item.value}
                secondary={{
                  data: [
                    {
                      type: item.trend > 0 ? 'increase' : 'decrease',
                      label: '环比',
                      value: `${Math.abs(item.trend)}%`,
                      color: item.trend > 0 ? 'danger' : 'success',
                    },
                  ],
                }}
                footer={
                  <div style={{ textAlign: 'right' }}>
                    <Button onClick={() => handleEdit(item)} style={{ marginRight: '8px' }}>
                      编辑标题
                    </Button>
                    <Button onClick={() => handleDelete(item.id)}>删除</Button>
                  </div>
                }
              />
            </Card>
          ))}
        </MetricsGroup>
      </div>
    </div>
  );
};

export const Layout = () => {
  const [enableGroup, setEnableGroup] = React.useState(true);

  const Comp = enableGroup ? MetricsGroup : React.Fragment;

  return (
    <div
      style={{
        width: '900px',
      }}
    >
      <div>
        <Button
          onClick={() => {
            setEnableGroup(!enableGroup);
          }}
        >
          {enableGroup ? '禁用分组' : '启用分组'}
        </Button>
      </div>
      <Comp>
        <PerformanceContent componentName="SeveralMetrics" count={100} itemRenderer={() => <SeveralMetrics />} />
      </Comp>
    </div>
  );
};

const SeveralMetrics = () => {
  const style = { width: '300px', display: 'inline-block' };
  const [value, setValue] = React.useState('¥10000000000000000000万');

  return (
    <>
      <Input value={value} onChange={e => setValue(e.currentTarget.value)} />
      <div
        style={{
          width: '900px',
          // display: 'flex',
          // flexWrap: 'wrap',
        }}
      >
        <Card style={style}>
          <Metrics
            mainTitle="销售金额销售金额销售金额销售金额销售金额销售金额销售金额销售金额销售金额销售金额销售金额销售金额销售金额销售金额销售金额销售金额销售金额销售金额销售金额销售金额销售金额销售金额销售金额销售金额销售金额销售金额销售金额销售金额"
            mainInfoIcon={{
              tooltip: {
                title: '默认启用字号缩小',
              },
            }}
            mainValue={value}
            secondary={{
              data: [
                {
                  type: 'increase',
                  label: '超超超超超超超长本月环比',
                  value: '17%',
                  color: 'danger',
                },
                {
                  type: 'decrease',
                  label: '上个周期',
                  value: '17%',
                  color: 'success',
                },
              ],
            }}
          />
        </Card>

        <Card style={style}>
          <Metrics
            mainTitle="销售金额"
            mainInfoIcon={{
              tooltip: {
                title: '默认启用字号缩小',
              },
            }}
            mainValue={value}
            secondary={{
              data: [
                {
                  type: 'increase',
                  label: '本月环比',
                  value: '17%',
                  color: 'danger',
                },
                {
                  type: 'decrease',
                  label: '上个周期',
                  value: '17%',
                  color: 'success',
                },
              ],
            }}
          />
        </Card>

        <Card style={style}>
          <Metrics
            mainTitle="销售金额"
            mainInfoIcon={{
              tooltip: {
                title: '默认启用字号缩小',
              },
            }}
            mainValue={value}
            secondary={{
              data: [
                {
                  type: 'increase',
                  label: '超超超超超超超长本月环比',
                  value: '17%',
                  color: 'danger',
                },
                {
                  type: 'decrease',
                  label: '上个周期',
                  value: '17%',
                  color: 'success',
                },
              ],
            }}
          />
        </Card>

        <Card style={style}>
          <Metrics
            mainTitle="销售金额"
            mainInfoIcon={{
              tooltip: {
                title: '默认启用字号缩小',
              },
            }}
            mainValue={value}
            secondary={{
              data: [
                {
                  type: 'increase',
                  label: '超超超超超超超长本月环比',
                  value: '17%',
                  color: 'danger',
                },
                {
                  type: 'decrease',
                  label: '上个周期',
                  value: '17%',
                  color: 'success',
                },
              ],
            }}
          />
        </Card>

        <Card style={style}>
          <Metrics
            mainTitle="销售金额"
            mainInfoIcon={{
              tooltip: {
                title: '默认启用字号缩小',
              },
            }}
            mainValue={value}
            secondary={{
              data: [
                {
                  type: 'increase',
                  label: '本月环比',
                  value: '17%',
                  color: 'danger',
                },
                {
                  type: 'decrease',
                  label: '上个周期',
                  value: '17%',
                  color: 'success',
                },
              ],
            }}
          />
        </Card>

        <Card style={style}>
          <Metrics
            mainTitle="销售金额"
            mainInfoIcon={{
              tooltip: {
                title: '默认启用字号缩小',
              },
            }}
            mainValue={value}
            secondary={{
              data: [
                {
                  type: 'increase',
                  label: '本月环比',
                  value: '17%',
                  color: 'danger',
                },
                {
                  type: 'decrease',
                  label: '上个周期',
                  value: '17%',
                  color: 'success',
                },
              ],
            }}
          />
        </Card>

        <Card style={style}>
          <Metrics
            mainTitle="销售金额"
            mainInfoIcon={{
              tooltip: {
                title: '默认启用字号缩小',
              },
            }}
            mainValue={value}
            secondary={{
              data: [
                {
                  type: 'increase',
                  label: '本月环比',
                  value: '17%',
                  color: 'danger',
                },
                {
                  type: 'decrease',
                  label: '上个周期',
                  value: '17%',
                  color: 'success',
                },
              ],
            }}
          />
        </Card>

        <Card style={style}>
          <Metrics
            mainTitle="销售金额"
            mainInfoIcon={{
              tooltip: {
                title: '默认启用字号缩小',
              },
            }}
            mainValue={value}
            secondary={{
              data: [
                {
                  type: 'increase',
                  label: '本月环比',
                  value: '17%',
                  color: 'danger',
                },
                {
                  type: 'decrease',
                  label: '上个周期',
                  value: '17%',
                  color: 'success',
                },
              ],
            }}
          />
        </Card>

        <Card style={style}>
          <Metrics
            mainTitle="销售金额"
            mainInfoIcon={{
              tooltip: {
                title: '默认启用字号缩小',
              },
            }}
            mainValue={value}
            secondary={{
              data: [
                {
                  type: 'increase',
                  label: '本月环比',
                  value: '17%',
                  color: 'danger',
                },
                {
                  type: 'decrease',
                  label: '上个周期',
                  value: '17%',
                  color: 'success',
                },
              ],
            }}
          />
        </Card>
      </div>
    </>
  );
};
