// @universe-design/react/src/metrics/stories/screenshot.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import CBasic from '../demo/basic';
import CChart from '../demo/chart';
import CFooter from '../demo/footer';
import CGroup from '../demo/group';
import CInfo from '../demo/info';
import CScale from '../demo/scale';
import CTitleContent from '../demo/title-content';
import CTrend from '../demo/trend';
import CWrap from '../demo/wrap';
if (!(CBasic as Story).storyName) {
  (CBasic as Story).storyName = 'Basic';
}
if (!(CChart as Story).storyName) {
  (CChart as Story).storyName = 'Chart';
}
if (!(CFooter as Story).storyName) {
  (CFooter as Story).storyName = 'Footer';
}
if (!(CGroup as Story).storyName) {
  (CGroup as Story).storyName = 'Group';
}
if (!(CInfo as Story).storyName) {
  (CInfo as Story).storyName = 'Info';
}
if (!(CScale as Story).storyName) {
  (CScale as Story).storyName = 'Scale';
}
if (!(CTitleContent as Story).storyName) {
  (CTitleContent as Story).storyName = 'Title content';
}
if (!(CTrend as Story).storyName) {
  (CTrend as Story).storyName = 'Trend';
}
if (!(CWrap as Story).storyName) {
  (CWrap as Story).storyName = 'Wrap';
}

export default {
  title: 'Components/Metrics/截图测试',
  tags: ['autodocs'],
};

export { CBasic, CChart, CFooter, CGroup, CInfo, CScale, CTitleContent, CTrend, CWrap };
