// @universe-design/react/es/shared/components/expand-button/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React from 'react';
import { ExpandableExpandDownFilled } from './ExpandDownFilled';
import { ExpandableDownBoldOutlined } from './DownBoldOutlined';
import clsx from 'clsx';
import { Button } from '../../../button';
import { useCssScope } from '../../../config-provider/hooks';
import { isFunction } from '../../../util';
import './index.less';
export const ExpandButton = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    isExpanded,
    icon,
    iconType,
    size = iconType === 'ExpandDownFilled' ? 'xxs' : 'small',
    onExpandChange,
    onClick,
    className,
    asPlaceholder = false,
    readonly = false,
    ...restProps
  } = _ref;
  const cssScope = useCssScope('expandButton');
  const handleClick = React.useCallback(e => {
    isFunction(onClick) && !readonly && onClick(e);
    isFunction(onExpandChange) && onExpandChange(!isExpanded);
  }, [isExpanded, onClick, onExpandChange, readonly]);
  const iconNode = (() => {
    if (isFunction(icon)) {
      return icon({
        isExpanded
      });
    }
    if (typeof icon !== 'undefined') {
      return icon;
    }
    if (iconType === 'ExpandDownFilled') {
      return /*#__PURE__*/React.createElement(ExpandableExpandDownFilled, {
        isExpanded: isExpanded
      });
    }
    if (iconType === 'DownBoldOutlined') {
      return /*#__PURE__*/React.createElement(ExpandableDownBoldOutlined, {
        isExpanded: isExpanded
      });
    }
  })();
  return /*#__PURE__*/React.createElement(Button, _extends({
    type: "icon",
    ref: ref,
    className: clsx(cssScope(), className, asPlaceholder && cssScope('-as-placeholder'), readonly && cssScope('-readonly'), size === 'xxs' && cssScope('-xxs')),
    icon: iconNode,
    onClick: handleClick,
    size: size === 'xxs' ? undefined : size
  }, restProps));
});
