// @universe-design/react/es/menu/Menu.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React, { useContext, useMemo } from 'react';
import clsx from 'clsx';
import RightBoldOutlined from '@universe-design/icons-react/RightBoldOutlined';
import RcMenu from './internal-menu';
import { SiderContext } from '../layout';
import { useClassPrefix, useCssScope } from '../config-provider/hooks';
import { MenuContext } from './MenuContext';
import { collapseMotion } from '../shared/motion';
import { milliSeconds2Seconds, isSubMenu } from './util';
import { devWarning, isFunction, isNonNullable, omit, warning } from '../util';
import { DropdownMenuContext } from '../dropdown/DropdownMenuContext';
import MenuMore from './MenuMore';
import { getPlacements } from './placements';
import { navMenuDefaults } from './constants';
import './style';

/**
 * @docgen
 */

const Menu = props => {
  const {
    mode: modeProp,
    inlineCollapsed = false,
    className,
    prefixCls: prefixClsInProps,
    expandIcon,
    textOverflow = modeProp !== 'normal',
    getPopupContainer = () => document.body,
    subMenuCloseDelay,
    subMenuOpenDelay,
    inlineIndent = navMenuDefaults.inlineIndent,
    showCheckmark = true,
    onClick: onClickProp,
    selectable: selectableProp,
    builtinPlacements: builtinPlacementsProp,
    overflowedIndicator: overflowedIndicatorProp,
    selectedKeys,
    defaultSelectedKeys
  } = props;
  const rootClasPrefix = useClassPrefix();
  const siderCtx = useContext(SiderContext);
  const dropdownCtx = useContext(DropdownMenuContext);
  /**
   * 对于menu的mode属性，首先采用用户定义的mode，再考虑:
   * 如果是dropdown中的menu，即存在dropdownCtx，则mode定义为"normal"
   * 如果是普通的menu，则mode定义为"vertical"
   */
  const mergedMode = React.useMemo(() =>
  // eslint-disable-next-line no-nested-ternary
  isNonNullable(modeProp) ? modeProp : dropdownCtx ? 'normal' : 'vertical', [dropdownCtx, modeProp]);
  const mergedPlacements = React.useMemo(() => {
    if (builtinPlacementsProp) {
      return builtinPlacementsProp;
    }
    return getPlacements(mergedMode);
  }, [builtinPlacementsProp, mergedMode]);
  /**
   * 对于menu的selectable属性，首先采用用户定义的selectable，再考虑:
   * 如果传递了 selectedKeys, 也认为 selectable 为 true
   * 如果是dropdown中的menu，即存在dropdownCtx，则selectable定义为false
   * 如果是普通的menu，则selectable定义为true
   */
  const mergedSelectable = React.useMemo(() => {
    if (isNonNullable(selectableProp)) {
      return selectableProp;
    }
    if (isNonNullable(selectedKeys) || isNonNullable(defaultSelectedKeys)) {
      return true;
    }
    return !dropdownCtx;
  }, [defaultSelectedKeys, dropdownCtx, selectableProp, selectedKeys]);
  const cssScope = useCssScope(mergedMode === 'normal' ? 'menu-normal' : 'menu');
  const prefixCls = prefixClsInProps || cssScope();
  process.env.NODE_ENV !== "production" ? warning(
  // eslint-disable-next-line react/destructuring-assignment
  !(siderCtx && props.inlineCollapsed !== undefined), devWarning('Menu', 'prop `inlineCollapsed` will be ignored inside a <Sider /> component')) : void 0;
  const mergedInlineCollapsed = siderCtx ? siderCtx.siderCollapsed : inlineCollapsed;
  const {
    children
  } = props;
  const includeSubMenu = React.Children.toArray(children).find(isSubMenu);
  const menuClassName = clsx(className, dropdownCtx === null || dropdownCtx === void 0 ? void 0 : dropdownCtx.className, {
    [`${prefixCls}-inline-collapsed`]: mergedInlineCollapsed,
    [`${prefixCls}-include-subMenu`]: includeSubMenu
  }, 'ud-scrollbar');
  const defaultMotions = {
    horizontal: {
      motionName: `${rootClasPrefix}-slide-up`
    },
    inline: collapseMotion(rootClasPrefix),
    other: {
      motionName: `${rootClasPrefix}-zoom-big`
    }
  };
  const rcMode = mergedMode === 'normal' ? 'vertical' : mergedMode;
  return /*#__PURE__*/React.createElement(MenuContext.Provider, {
    value: useMemo(() => ({
      prefixCls,
      inlineCollapsed: mergedInlineCollapsed,
      mode: mergedMode,
      selectable: mergedSelectable,
      textOverflow,
      firstLevel: true,
      showCheckmark
    }), [mergedInlineCollapsed, mergedMode, mergedSelectable, prefixCls, showCheckmark, textOverflow])
  }, /*#__PURE__*/React.createElement(RcMenu, _extends({
    builtinPlacements: mergedPlacements,
    getPopupContainer: getPopupContainer
  }, omit(props, ['showCheckmark', 'textOverflow']), {
    overflowedIndicator: overflowedIndicatorProp !== null && overflowedIndicatorProp !== void 0 ? overflowedIndicatorProp : /*#__PURE__*/React.createElement(MenuMore, null),
    onClick: React.useCallback(e => {
      isFunction(onClickProp) && onClickProp(e);
      isFunction(dropdownCtx === null || dropdownCtx === void 0 ? void 0 : dropdownCtx.onClick) && (dropdownCtx === null || dropdownCtx === void 0 ? void 0 : dropdownCtx.onClick(e));
    }, [dropdownCtx, onClickProp]),
    mode: rcMode,
    selectable: mergedSelectable,
    inlineCollapsed: mergedInlineCollapsed,
    className: menuClassName,
    prefixCls: prefixCls,
    direction: "ltr",
    expandIcon: () => /*#__PURE__*/React.createElement("span", {
      className: `${prefixCls}-submenu-arrow`
    }, expandIcon || /*#__PURE__*/React.createElement(RightBoldOutlined, {
      className: `${prefixCls}-submenu-arrow-icon`
    })),
    defaultMotions: defaultMotions
  }, milliSeconds2Seconds({
    subMenuOpenDelay,
    subMenuCloseDelay
  }), {
    inlineIndent: inlineIndent
  })));
};
export default Menu;
