// @universe-design/react/es/menu/MenuTitle.js
//.u1qumt79{line-height:inherit;}
import u1wwc0yi from "./../.stylex/u1wwc0yi";
//.uf5ikls{font-weight:inherit;}
import u12n1ij7 from "./../.stylex/u12n1ij7";
//.ujoy568{font-size:inherit;}
import u1g5jyy from "./../.stylex/u1g5jyy";
//.uqerorx{flex-grow:1;}
import u99gbyh from "./../.stylex/u99gbyh";
//.u10tiqix{min-width:0px;}
import urmav4p from "./../.stylex/urmav4p";
import stylex from '@universe-design/stylex';
import clsx from 'clsx';
import React from 'react';
import { Text } from '../text';
import { MenuContext } from './MenuContext';
import { RcMenuContext } from './internal-menu/context/MenuContext';
const sxMenuTitleContent = {
  "props": {
    "Bg96gwp": u1wwc0yi,
    "Bhrd7zp": u12n1ij7,
    "Be2twd7": u1g5jyy,
    "Bh6795r": u99gbyh,
    "Bf4jedk": urmav4p
  }
};
export const MenuTitle = _ref => {
  let {
    showEllipsis,
    title,
    style
  } = _ref;
  const {
    mode: rcMode
  } = React.useContext(RcMenuContext);
  const {
    prefixCls
  } = React.useContext(MenuContext);
  if (showEllipsis) {
    const lineClamp = rcMode === 'vertical' || rcMode === 'inline' ? 2 : 1;
    return /*#__PURE__*/React.createElement(Text, {
      as: "div",
      className: clsx(`${prefixCls}-item-title-content`, `${prefixCls}-item-title-content-text-overflow`),
      style: style,
      sx: [sxMenuTitleContent],
      lineClamp: lineClamp
    }, title);
  }
  return /*#__PURE__*/React.createElement(Text, {
    as: "div",
    style: style,
    className: `${prefixCls}-item-title-content`,
    sx: [sxMenuTitleContent]
  }, title);
};
