// @universe-design/react/es/menu/MenuItem.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.u64fuq3{width:20px;}
import uevmt6d from "./../.stylex/uevmt6d";
//.ui64zpg{flex-shrink:0;}
import uidkx17 from "./../.stylex/uidkx17";
import * as React from 'react';
import ListCheckBoldOutlined from '@universe-design/icons-react/ListCheckBoldOutlined';
import { RcMenuItem } from './internal-menu';
import clsx from 'clsx';
import stylex from '@universe-design/stylex';
import { cssTypographyBody0 } from '../stylex';
import { MenuContext } from './MenuContext';
import { Tooltip } from '../tooltip';
import { toFlatArray } from '../util';
import { HorizontalMenuInk } from './HorizontalMenuInk';
import { MenuTitle } from './MenuTitle';
import { EmptyElement } from '../shared/components/FirstLineAlign';
import './style';
/**
 * @docgen
 * @public
 */

const MenuItemSuffixIcon = _ref => {
  let {
    isSelected
  } = _ref;
  const {
    prefixCls,
    mode: rootMode,
    inlineCollapsed,
    showCheckmark,
    selectable
  } = React.useContext(MenuContext);
  return selectable && showCheckmark && rootMode === 'normal' && !inlineCollapsed ? /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("span", {
    className: stylex({
      "props": {
        "a9b677": uevmt6d,
        "Bnnss6s": uidkx17
      }
    })
  }), isSelected && /*#__PURE__*/React.createElement(ListCheckBoldOutlined, {
    className: `${prefixCls}-item-suffix-icon`
  })) : null;
};
const MenuItem = props => {
  const {
    className,
    children,
    title,
    icon,
    textOverflow: textOverflowInProps,
    disabled,
    ...rest
  } = props;
  const {
    prefixCls,
    inlineCollapsed,
    textOverflow,
    firstLevel,
    mode: rootMode
  } = React.useContext(MenuContext);
  let tooltipTitle = title;
  if (typeof title === 'undefined') {
    tooltipTitle = firstLevel ? children : '';
  } else if (title === false) {
    tooltipTitle = '';
  }
  const tooltipProps = {
    title: tooltipTitle
  };
  if (!inlineCollapsed) {
    // tooltipProps.title = null; // 这个属性变化会导致 children remount，从而引起 children transition 动画失效
    // Reset `visible` to fix control mode tooltip display not correct
    tooltipProps.visible = false;
  }
  const childrenLength = toFlatArray(children).length;
  const cls = clsx(className, `${prefixCls}-item--root-${rootMode}`, {
    [`${prefixCls}-item-only-child`]: (icon ? childrenLength + 1 : childrenLength) === 1
  }, stylex(rootMode === 'normal' && cssTypographyBody0));
  return /*#__PURE__*/React.createElement(Tooltip, _extends({}, tooltipProps, {
    overlayClassName: `${prefixCls}-inline-collapsed-tooltip`,
    placement: "right"
  }), /*#__PURE__*/React.createElement(RcMenuItem, _extends({}, rest, {
    className: cls,
    title: typeof title === 'string' ? title : undefined,
    itemIcon: MenuItemSuffixIcon,
    disabled: disabled
  }), (() => icon && /*#__PURE__*/React.createElement("div", {
    className: clsx(`${prefixCls}-icon-wrap`, disabled && `${prefixCls}-icon-wrap-disabled`)
  }, icon))(), (() => {
    const collapsedItem = inlineCollapsed && firstLevel;
    // 收起模式下显示 icon 或是 children 的第一个字符
    if (collapsedItem) {
      if (!icon && typeof children === 'string') {
        return /*#__PURE__*/React.createElement("div", {
          className: `${prefixCls}-inline-collapsed-noicon`
        }, children.charAt(0));
      }
    }
    const showEllipsis = !collapsedItem && (textOverflowInProps !== null && textOverflowInProps !== void 0 ? textOverflowInProps : textOverflow);
    return /*#__PURE__*/React.createElement(MenuTitle, {
      style: collapsedItem ? {
        width: 0,
        overflow: 'hidden'
      } : {},
      showEllipsis: showEllipsis,
      title: children
    });
  })(), /*#__PURE__*/React.createElement(EmptyElement, null), /*#__PURE__*/React.createElement(HorizontalMenuInk, null)));
};
if (process.env.NODE_ENV !== "production") {
  MenuItem.displayName = 'MenuItem';
}
export default MenuItem;
