// @universe-design/react/es/menu/placements.js
export function getOffset(mode) {
  let popupOffset;
  switch (mode) {
    case 'horizontal':
      {
        // 设计规范要 x 方向刚好压着边（border 重合），y 方向 menu item 要和 sub menu item 的第一个对齐

        popupOffset = [0, -(8 + 1 + 1)];
        break;
      }
    case 'vertical':
    case 'inline':
      {
        // 设计规范要 x 方向刚好压着边（border 重合），
        // y 方向 menu item 要和 sub menu item 的第一个对齐

        popupOffset = [8, -(8 + 1 + 1)];
      }
      break;
    case 'normal':
      {
        popupOffset = [3, -(2 + 1 + 1)];
      }
      break;
    default:
      break;
  }
  return popupOffset;
}
export function getPlacements(mode) {
  const popupOffset = getOffset(mode);
  const defaultAlign = {
    targetOffset: [0, 0],
    ignoreShake: true,
    overflow: {
      adjustX: 1,
      adjustY: 1,
      alwaysByViewport: true
    }
  };
  return {
    topLeft: {
      points: ['bl', 'tl'],
      offset: popupOffset,
      ...defaultAlign
    },
    bottomLeft: {
      points: ['tl', 'bl'],
      offset: popupOffset,
      ...defaultAlign
    },
    leftTop: {
      points: ['tr', 'tl'],
      offset: popupOffset,
      ...defaultAlign
    },
    rightTop: {
      points: ['tl', 'tr'],
      offset: popupOffset,
      ...defaultAlign
    },
    left: {
      points: ['cr', 'cl'],
      offset: popupOffset,
      ...defaultAlign
    },
    right: {
      points: ['cl', 'cr'],
      offset: popupOffset,
      ...defaultAlign
    },
    top: {
      points: ['bc', 'tc'],
      offset: popupOffset,
      ...defaultAlign
    },
    bottom: {
      points: ['tc', 'bc'],
      offset: popupOffset,
      ...defaultAlign
    },
    topRight: {
      points: ['br', 'tr'],
      offset: popupOffset,
      ...defaultAlign
    },
    bottomRight: {
      points: ['tr', 'br'],
      offset: popupOffset,
      ...defaultAlign
    },
    rightBottom: {
      points: ['bl', 'br'],
      offset: popupOffset,
      ...defaultAlign
    },
    leftBottom: {
      points: ['br', 'bl'],
      offset: popupOffset,
      ...defaultAlign
    }
  };
}
