// @universe-design/styles/ts/ask/index.light.atom.ts
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
export const asktoken_knowledge_btn_bg_copying_hover = '#fff1d1';
export const asktoken_knowledge_btn_bg_copying_normal = '#fff7e5';
export const asktoken_knowledge_btn_bg_downloading_hover = '#ddf9cb';
export const asktoken_knowledge_btn_bg_downloading_normal = '#ebfae1';
export const asktoken_knowledge_example_card_bg =
  'linear-gradient(to bottom, rgba(240, 244, 255, 0.4) 30%, rgba(240, 244, 255, 0) 100%)';
export const asktoken_knowledge_example_card_bg_2 =
  'linear-gradient(to bottom, #ffffff 30%, rgba(255, 255, 255, 0) 100%)';
export const asktoken_knowledge_fill_file_blue = '#f0f4ff';
export const asktoken_knowledge_fill_file_green = '#e4fae1';
export const asktoken_knowledge_fill_file_indigo = '#f2f3fd';
export const asktoken_knowledge_fill_file_neutral = '#f5f6f7';
export const asktoken_knowledge_fill_file_orange = '#fff3e5';
export const asktoken_knowledge_fill_file_red = '#fef0f0';
export const asktoken_knowledge_onboarding_text_gradient = 'linear-gradient(to bottom, #4752e6 0%, #8d55ed 100%)';
export const asktoken_knowledge_rank_icon_bg = '#ffad1f';
export const asktoken_knowledge_rank_icon_outline = '#ffe880';
export const asktoken_knowledge_scope_bg = '#ffffff';
