// @universe-design/react/src/checkbox/Checkbox.tsx
import type { Ref, ChangeEvent, CSSProperties, MouseEventHandler } from 'react';
import React, { useRef, forwardRef, useImperativeHandle } from 'react';
import clsx from 'clsx';
import stylex from '@universe-design/stylex';
import { useCheckbox } from '@universe-design/react-utilities';
import { useClassPrefix, useCssScope } from '../config-provider/hooks';
import { CheckboxCheckedSvg, CheckboxIndeterminateSvg } from './_assets/svgs';
import { useCheckboxGroupContext } from './useCheckboxContext';
import { isFunction, omit } from '../util';
import type { CheckBoxChangeEvent, CheckBoxValue } from './interface';
import { markAsDisabledableComp } from '../shared/markAsDisabledableComp';
import { cssTypographyBody0 } from '../stylex';

import './style';

// feat 精细的布局，所有样式基于 input dome 的真实状态，具有最高的规范性和一致性
type HTMLInputPropsType = React.InputHTMLAttributes<HTMLInputElement>;
export interface CheckboxProps
  extends Omit<HTMLInputPropsType, 'onChange' | 'onMouseEnter' | 'onMouseLeave' | 'value'> {
  /**
   * `className`属性透传
   */
  className?: string;
  /**
   * `style`属性透传
   */
  style?: CSSProperties;
  /**
   * 初始是否选中
   * @defaultValue false
   */
  defaultChecked?: boolean;
  /**
   * 指定当前是否选中
   */
  checked?: boolean;
  /**
   * 禁用
   * @defaultValue false
   */
  disabled?: boolean;
  /**
   * 错误状态
   * @defaultValue false
   */
  error?: boolean;
  /**
   * 设置`indeterminate`状态，只负责样式控制
   * @defaultValue false
   */
  indeterminate?: boolean;
  /**
   * `name`属性透传
   * @defaultValue false
   */
  name?: string;
  /**
   * 根据`value`进行比较，判断是否选中
   */
  value?: CheckBoxValue;
  /**
   * 自动获取焦点
   * @defaultValue false
   */
  autoFocus?: boolean;
  /**
   * 变化时回调函数
   */
  onChange?: (e: CheckBoxChangeEvent) => void;
  /**
   * 鼠标移入事件
   */
  onMouseEnter?: MouseEventHandler<HTMLLabelElement>;
  /**
   * 鼠标移出事件
   */
  onMouseLeave?: MouseEventHandler<HTMLLabelElement>;
  /**
   * 回调返回 value 作为参数
   */
  onCheckChange?: (checked: boolean) => void;
}

const CheckboxComp = forwardRef((props: CheckboxProps, ref: Ref<HTMLInputElement>) => {
  const {
    onChange,
    onCheckChange,
    onMouseEnter,
    onMouseLeave,
    className,
    children,
    value,
    disabled,
    name,
    error,
    ...otherProps
  } = props;

  const inputRef = useRef<HTMLInputElement>(null);
  const prefix = useClassPrefix();
  const cssScope = useCssScope('checkbox');
  const context = useCheckboxGroupContext();
  const checkboxProps = {
    ...otherProps,
    className: clsx(stylex(cssTypographyBody0), className),
    disabled,
    name,
    error,
    classPrefix: `${prefix}__checkbox`,
    onChange: handleChange,
  };

  if (context) {
    checkboxProps.disabled = disabled || context.disabled;
    checkboxProps.name = context.name;
    checkboxProps.error = error || context.error;
    checkboxProps.checked = Boolean(context.value?.includes(value));
  } else if ('checked' in props) {
    checkboxProps.checked = Boolean(props.checked);
  }

  function handleChange(checked: boolean, e: ChangeEvent<HTMLInputElement>): void {
    const eventParams: CheckBoxChangeEvent = {
      target: {
        type: 'checkbox',
        checked,
        value,
      },
      stopPropagation() {
        e.stopPropagation();
      },
      preventDefault() {
        e.preventDefault();
      },
      nativeEvent: e.nativeEvent,
      reactEvent: e,
    };

    if (context?.onChange) {
      context.onChange(eventParams);
    }
    if (isFunction(onChange)) {
      onChange(eventParams);
    }
    if (isFunction(onCheckChange)) {
      onCheckChange(checked);
    }
  }

  const { checked, indeterminate, inputProps, descriptionProps, wrapperProps } = useCheckbox(checkboxProps, inputRef);

  useImperativeHandle<HTMLInputElement | null, HTMLInputElement | null>(ref, () => inputRef.current);

  /**
   * 由于 background 有渐变动画效果，但 checked-svg 与 indeterminate-svg 的切换是瞬间执行的，所以在切换的瞬间会出现只有 background 没有对勾或横杠的闪黑现象
   * 修改方案是: 同时返回 checked-svg 与 indeterminate-svg，而不是让两者做切换。在 checked 状态下将 checked-svg 的透明度置为 1 而 indeterminate-svg 的透明度置为 0；indeterminate 状态下同理
   */
  return (
    <label
      {...wrapperProps}
      className={wrapperProps.className}
      onMouseEnter={onMouseEnter}
      onMouseLeave={onMouseLeave}
      data-testid="ud__checkbox"
    >
      <span className={cssScope()}>
        <input
          {...omit(otherProps, ['error', 'className', 'style', 'defaultChecked', 'indeterminate'])}
          {...inputProps}
          ref={inputRef}
        />
        <span className={cssScope('__wallpaper')}>
          <CheckboxCheckedSvg
            className={clsx(cssScope('__checked-svg'), checked && !indeterminate && cssScope('__checked-svg-shown'))}
          />
          <CheckboxIndeterminateSvg
            className={clsx(cssScope('__indeterminate-svg'), indeterminate && cssScope('__indeterminate-svg-shown'))}
          />
        </span>
      </span>
      {children ? <span {...descriptionProps}>{children}</span> : null}
    </label>
  );
});

type CheckboxType = typeof CheckboxComp & { __UD_CHECKBOX: boolean };

export const Checkbox = /* #__PURE__*/ markAsDisabledableComp(CheckboxComp as CheckboxType);

Checkbox.__UD_CHECKBOX = true;
if (__DEV__) {
  Checkbox.displayName = 'Checkbox';
}
