// @universe-design/react/src/checkbox/Group.tsx
import clsx from 'clsx';
import type { ChangeEvent, CSSProperties } from 'react';
import React, { forwardRef } from 'react';
import { useControlledState } from '@universe-design/react-utilities';
import { useCssScope } from '../config-provider/hooks';
import { CheckboxGroupContextProvider } from './context';
import type { CheckBoxChangeEvent, CheckBoxValue } from './interface';
import { Checkbox } from './Checkbox';
import { isFunction } from '../util';

import './style';

export interface CheckboxOption {
  /**
   * 复选框标签
   */
  label: React.ReactNode;
  /**
   * 根据`value`进行比较，判断是否选中
   */
  value: CheckBoxValue;
  /**
   * `style`属性透传
   */
  style?: React.CSSProperties;
  /**
   * `className` 属性透传
   */
  className?: string;
  /**
   * 禁用
   * @defaultValue false
   */
  disabled?: boolean;
  /**
   * 错误状态
   * @defaultValue false
   */
  error?: boolean;
  /**
   * 变化时回调函数
   */
  onChange?: (e: ChangeEvent<HTMLInputElement>) => void;
}

export interface CheckboxGroupProps {
  /**
   * 复选框组下所有`input[type="checkbox"]`的`name`属性
   */
  name?: string;
  /**
   * `className`属性透传
   */
  className?: string;
  /**
   * `style`属性透传
   */
  style?: CSSProperties;
  /**
   * 整组失效
   * @defaultValue false
   */
  disabled?: boolean;
  /**
   * 错误状态
   * @defaultValue false
   */
  error?: boolean;
  /**
   * 指定选中的选项
   */
  value?: CheckBoxValue[];
  /**
   * 默认选中的选项
   * @defaultValue []
   */
  defaultValue?: CheckBoxValue[];
  /**
   * 指定可选项
   * @defaultValue []
   */
  options?: Array<CheckboxOption | string>;
  /**
   * 子组件，即 CheckBox
   */
  children?: React.ReactNode;
  /**
   * 变化时回调函数
   */
  onChange?: (checkedValue: CheckBoxValue[]) => void;
  /**
   * 鼠标移入事件
   */
  onMouseEnter?: React.MouseEventHandler<HTMLDivElement>;
  /**
   * 鼠标移出事件
   */
  onMouseLeave?: React.MouseEventHandler<HTMLDivElement>;
}

const CheckboxGroup = forwardRef<HTMLDivElement, CheckboxGroupProps>((props, ref) => {
  const groupCssScope = useCssScope('checkbox-group');
  const [value, setValue] = useControlledState(props.defaultValue, props.value ?? [], {
    isControlledOverride: 'value' in props,
  });

  const triggerOnChange = (data: CheckBoxValue[]) => {
    isFunction(props.onChange) && props.onChange(data);
    !('value' in props) && setValue(data);
  };

  const onCheckedChange = (e: CheckBoxChangeEvent) => {
    const { value: changeValue } = e.target;
    // 在 group 模式下，如果没有提供 val，不会触发 onChange。
    if (changeValue === undefined) {
      return;
    }
    if (value) {
      const result = [...value];
      const index = value.indexOf(changeValue);
      if (index === -1) {
        result.push(changeValue);
      } else {
        result.splice(index, 1);
      }
      triggerOnChange(result);
    } else {
      triggerOnChange([changeValue]);
    }
  };

  const isChecked = (data: CheckBoxValue) => value?.includes(data);

  const renderGroup = () => {
    const { className, options, disabled, children, style, error, onMouseEnter, onMouseLeave } = props;
    let renderCheckbox = children;
    if (options && options.length > 0) {
      renderCheckbox = options.map(option => {
        if (typeof option === 'string') {
          return (
            <Checkbox key={option} disabled={disabled} value={option} checked={isChecked(option)} error={error}>
              {option}
            </Checkbox>
          );
        }
        return (
          <Checkbox
            key={`checkobx-option-${option.value}`}
            disabled={option.disabled || disabled}
            value={option.value}
            checked={isChecked(option.value)}
            style={option.style}
            className={option.className}
            error={option.error || error}
          >
            {option.label}
          </Checkbox>
        );
      });
    }

    const classNames = clsx(groupCssScope(), className);
    return (
      <div className={classNames} style={style} onMouseEnter={onMouseEnter} onMouseLeave={onMouseLeave} ref={ref}>
        {renderCheckbox}
      </div>
    );
  };

  return (
    <CheckboxGroupContextProvider
      value={{
        onChange: onCheckedChange,
        value,
        name: props.name,
        disabled: props.disabled,
        error: props.error,
      }}
    >
      {renderGroup()}
    </CheckboxGroupContextProvider>
  );
});

export default React.memo(CheckboxGroup);
