// @universe-design/react/es/input/Group.js
import * as React from 'react';
import clsx from 'clsx';
import { getDefaultSizeAbbr } from '../util';
import { useCssScope } from '../config-provider/hooks';
import { InputGroupSizeContextProvider } from '../config-provider/size';
import './style';

/**
 * 组合输入框： `Input.Group`
 */

export const Group = props => {
  const {
    style,
    className,
    size = 'medium',
    compact,
    children,
    onMouseEnter,
    onMouseLeave,
    onFocus,
    onBlur
  } = props;
  const cssScope = useCssScope('input-group');
  const usedSize = getDefaultSizeAbbr(size, 'medium');
  const cls = clsx(cssScope(), {
    [cssScope(`--size-${usedSize}`)]: usedSize,
    [cssScope('--compact')]: compact
  }, className);
  return /*#__PURE__*/React.createElement(InputGroupSizeContextProvider, {
    size: size
  }, /*#__PURE__*/React.createElement("span", {
    className: cls,
    style: style,
    onMouseEnter: onMouseEnter,
    onMouseLeave: onMouseLeave,
    onFocus: onFocus,
    onBlur: onBlur
  }, children));
};
export default Group;
