// @universe-design/react/es/input/Input.js
import _extends from "@babel/runtime/helpers/esm/extends";
//.u11u7vat{min-width:20px;}
import u1n2z6e7 from "./../.stylex/u1n2z6e7";
//.u1euv43f{position:absolute;}
import u1u1p7mj from "./../.stylex/u1u1p7mj";
//.u1xynx9j{left:0px;}
import u184z4yl from "./../.stylex/u184z4yl";
//.u113wtx2{top:0px;}
import u1981n0b from "./../.stylex/u1981n0b";
//.u5gq2j6{bottom:0px;}
import u8rtfz1 from "./../.stylex/u8rtfz1";
//.u6dzj5z{max-width:100%;}
import u9hwb1v from "./../.stylex/u9hwb1v";
//.ud7fpy0{visibility:hidden;}
import u1xau0jj from "./../.stylex/u1xau0jj";
//.u10pi13n{position:relative;}
import u5jxv71 from "./../.stylex/u5jxv71";
//.uqerorx{flex-grow:1;}
import u99gbyh from "./../.stylex/u99gbyh";
//.u10tiqix{min-width:0px;}
import urmav4p from "./../.stylex/urmav4p";
//.u22iagw{display:flex;}
import ukchsq7 from "./../.stylex/ukchsq7";
//.u122n59{align-items:center;}
import ubyhozq from "./../.stylex/ubyhozq";
//.uly5x3f{width:100%;}
import u13d3rh8 from "./../.stylex/u13d3rh8";
/** @jsxRuntime classic */
/** @jsx jsx */
/** @jsxFrag React.Fragment */
import React, { forwardRef, useRef, useState, useImperativeHandle, useCallback, useEffect } from 'react';
import clsx from 'clsx';
import EmptyFilled from '@universe-design/icons-react/EmptyFilled';
import stylex from '@universe-design/stylex';
import { sxOverflowEllipsis } from '../stylex';
import { getDefaultSizeAbbr } from '../util/size';
import { isFunction } from '../util/assert';
import { focus } from '../util';
import { useComponentSize, useInputGroupSize, useCssScope } from '../config-provider/hooks';
import NativeInput from '../shared/components/Input';
import CounterComp from './components/Counter';
import { useInput } from '@universe-design/react-input';
import { FadeTransition } from '../shared/components/transition/FadeTransition';
import { useMouseOver } from '../hooks';
import { resolveOnChange } from './utils';
import { eventContains, isOverflowed } from '../util/dom';
import { useTextOverflowPopperSidecar } from '../config-provider/hooks/useTextOverflowPopperSidecar';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { jsx } from '@universe-design/stylex/react';
import './style';
export const Input = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    style,
    prefixCls,
    className,
    readOnly,
    placeholder,
    disabled,
    size,
    prefix,
    suffix,
    addonBefore,
    addonAfter,
    defaultValue = '',
    value: propsValue,
    allowClear,
    type,
    error,
    onChange,
    onKeyDown,
    onPressEnter,
    onClear,
    counter: counterProp,
    slicer,
    maxLength: maxLengthProp,
    showCount,
    truncateExcessWords,
    onCompositionStart: onCompositionStartProp,
    onCompositionEnd: onCompositionEndProp,
    sx,
    ...otherProps
  } = props;
  const cssScope = useCssScope('input');
  const defaultSize = useComponentSize();
  const inputGroupSize = useInputGroupSize();
  const [focused, setFocused] = useState(false);
  const inputRef = useRef(null);
  const timeout = useRef();
  const {
    isMouseOver,
    handleMouseEnter,
    handleMouseLeave
  } = useMouseOver();
  const currentSize = getDefaultSizeAbbr(size || inputGroupSize || defaultSize);
  const tagSize = currentSize === 'lg' ? 'medium' : 'small';
  const handleInputChange = useCallback((ref, event, value) => resolveOnChange(ref, event, onChange, value), [onChange]);
  const {
    nativeMaxLength,
    value,
    setValue,
    count,
    onCompositionEnd,
    onCompositionStart,
    internalOnChange
  } = useInput({
    maxLength: maxLengthProp,
    showCount,
    truncateExcessWords,
    counter: counterProp,
    slicer,
    defaultValue,
    inputRef,
    onChange: handleInputChange,
    onCompositionStart: onCompositionStartProp,
    onCompositionEnd: onCompositionEndProp,
    focused,
    ...('value' in props ? {
      value: propsValue
    } : {})
  });
  const placeholderRef = React.useRef(null);
  const stop = useCallback(() => {
    timeout.current && clearTimeout(timeout.current);
  }, []);
  const start = useCallback(() => {
    timeout.current && clearTimeout(timeout.current);
    timeout.current = setTimeout(() => {
      if (inputRef.current && inputRef.current.getAttribute('type') === 'password' && inputRef.current.getAttribute('value')) {
        inputRef.current.removeAttribute('value');
      }
      // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    }, 13);
  }, []);
  useEffect(() => {
    stop();
    start();
    return () => stop();
  }, [value, start, stop, type]);
  useImperativeHandle(ref, () => inputRef.current);
  const renderAddonBefore = cls => {
    if (addonBefore) {
      return jsx("span", {
        className: `${cls}__addon-before`
      }, addonBefore);
    }
    return null;
  };
  const renderAddonAfter = cls => {
    if (addonAfter) {
      return jsx("span", {
        className: `${cls}__addon-after`
      }, addonAfter);
    }
    return null;
  };
  const renderPrefix = cls => {
    if (prefix) {
      return jsx("span", {
        className: `${cls}__prefix`
      }, prefix);
    }
    return null;
  };
  const renderSuffix = cls => {
    if (suffix || showCount) {
      return jsx("span", {
        className: `${cls}__suffix`
      }, suffix, jsx(CounterComp
      // marginLeft 会根据 input Size 变化吗
      , {
        style: {
          marginLeft: suffix ? 8 : 0
        },
        showCount: showCount,
        count: count,
        maxLength: maxLengthProp,
        tagSize: tagSize
      }));
    }
    return null;
  };
  const onFocusChange = useCallback(isFocus => {
    stop();
    start();
    setFocused(isFocus);
  }, [start, stop]);
  const handleClear = e => {
    var _inputRef$current;
    if (!('value' in props)) {
      setValue('');
    }
    if (inputRef.current) {
      resolveOnChange(inputRef.current, e, onChange, '');
    }
    (_inputRef$current = inputRef.current) === null || _inputRef$current === void 0 ? void 0 : _inputRef$current.focus();
    isFunction(onClear) && onClear();
  };
  const handleKeyDown = e => {
    /* istanbul ignore else  */
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    if (e.keyCode === 13 && onPressEnter) {
      onPressEnter(e);
    }
    /* istanbul ignore else  */
    if (onKeyDown) {
      onKeyDown(e);
    }
  };
  const renderClearIcon = cls => {
    if (allowClear && !disabled) {
      // 输入框激活态非 hover 状态不展示 query 清除按钮
      const isShow = isMouseOver && Boolean(value);
      const clx = clsx({
        [`${cls}__clear-icon`]: true,
        [`${cls}__clear-icon-show`]: isShow
      });
      return jsx(FadeTransition, {
        in: isShow,
        timeout: 400
      }, jsx("span", {
        className: clx,
        onClick: handleClear
      }, jsx(EmptyFilled, null)));
    }
    return null;
  };
  const TextOverflowPopper = useTextOverflowPopperSidecar();
  // 当value 为空时才应该显示 tooltip
  const isPlaceholderVisible = Boolean(!value && placeholder);
  const [isPlaceholderTooltipVisible, setIsPlaceholderTooltipVisible] = React.useState(false);
  return (() => {
    const cls = prefixCls || cssScope();
    const rootCls = clsx({
      [cls]: true,
      [`${cls}--error`]: error,
      [`${cls}-addon`]: addonAfter || addonBefore,
      [`${cls}--focused`]: focused,
      [`${cls}--disabled`]: disabled,
      [`${cls}--size-${currentSize}`]: currentSize
    }, className, stylex({
      "props": {
        "a9b677": u13d3rh8
      }
    }, ...(sx !== null && sx !== void 0 ? sx : [])));
    return jsx("div", {
      style: style,
      className: rootCls
    }, renderAddonBefore(cls), jsx("label", {
      // 用 label 是为了嵌套 label 情况下能正常使用
      className: `${cls}-input-wrap`,
      onMouseEnter: handleMouseEnter,
      onMouseLeave: handleMouseLeave,
      onMouseDown: ev => {
        if (ev.target === inputRef.current) {
          // 如果是点击 input 的话就不需要做任何操作
          return;
        }
        // 如果事件是 wrap 里面的元素
        if (eventContains(ev.currentTarget, ev.nativeEvent)) {
          // 点击 wrap 内部元素不会让 input 失焦（mousedown 的 blur 逻辑会在这段同步代码之后执行）
          ev.preventDefault();
          // 点击 wrap 内部元素 focus 到 input 上
          focus(inputRef, {
            selectTextIfInput: false
          });
        }
      },
      onClick: ev => {
        // 如果是 wrap 内部元素触发的 onClick，我们禁用 label 默认行为，因为 input 会触发 onClick，避免 bubble 两次（有对 prefix 和 suffix 误伤的概率）。
        if (eventContains(ev.currentTarget, ev.nativeEvent)) {
          ev.preventDefault();
        }
      }
    }, renderPrefix(cls), jsx(TextOverflowPopper, {
      title: jsx(React.Fragment, null, placeholder),
      disabled: disabled,
      checkOverflow: () => {
        if (placeholderRef.current) {
          return isOverflowed(placeholderRef.current);
        }
        return false;
      },
      visible: isPlaceholderTooltipVisible && isPlaceholderVisible,
      onVisibleChange: visible => {
        setIsPlaceholderTooltipVisible(visible);
      }
    }, jsx("div", {
      className: clsx(`${cls}-input__placeholder__wrapper`, stylex({
        "props": {
          "qhf8xq": u5jxv71,
          "Bh6795r": u99gbyh,
          "Bf4jedk": urmav4p,
          "mc9l5x": ukchsq7,
          "Bt984gj": ubyhozq
        }
      })),
      onMouseLeave: () => {
        /**
         * 在 mouseleave 时，可能不会触发 tooltip 的 visibleChange
         * 所以这里必须手动重置一下
         */
        setIsPlaceholderTooltipVisible(false);
      }
    }, jsx("div", {
      ref: placeholderRef,
      className: clsx(`${cls}-input__placeholder`, stylex(sxOverflowEllipsis, {
        "props": {
          "qhf8xq": u1u1p7mj,
          "oyh7mz": u184z4yl,
          "Bhzewxz": u1981n0b,
          "B5kzvoi": u8rtfz1,
          "B2u0y6b": u9hwb1v,
          "Bcdw1i0": u1xau0jj
        }
      }))
    }, placeholder), jsx(NativeInput, _extends({}, otherProps, {
      type: type,
      ref: inputRef,
      readOnly: readOnly,
      disabled: disabled,
      value: value,
      onChange: internalOnChange,
      onKeyDown: handleKeyDown,
      onFocusChange: onFocusChange,
      maxLength: nativeMaxLength,
      onCompositionEnd: onCompositionEnd,
      onCompositionStart: onCompositionStart,
      placeholder: placeholder,
      sx: [{
        "props": {
          "a9b677": u13d3rh8,
          "Bf4jedk": u1n2z6e7
        }
      }]
    })))), renderClearIcon(cls), renderSuffix(cls)), renderAddonAfter(cls));
  })();
});
if (process.env.NODE_ENV !== "production") {
  Input.displayName = 'Input';
}
