// @universe-design/react/es/input/Password.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React, { useState } from 'react';
import { Input } from './Input';
import VisibleOutlined from '@universe-design/icons-react/VisibleOutlined';
import VisibleLockOutlined from '@universe-design/icons-react/VisibleLockOutlined';
import { useCssScope } from '../config-provider/hooks';
import { omit } from '../util';
import { Button } from '../button';
import './style';

/**
 * 用于密码输入框 `Input.Password`
 */

const ActionMap = {
  click: 'onClick',
  hover: 'onMouseOver'
};
export const Password = /*#__PURE__*/React.forwardRef((props, ref) => {
  const {
    action = 'click',
    size,
    visibilityToggle = true,
    iconRender = visible => visible ? /*#__PURE__*/React.createElement(Button, {
      type: "icon",
      size: "small",
      icon: /*#__PURE__*/React.createElement(VisibleOutlined, null)
    }) : /*#__PURE__*/React.createElement(Button, {
      type: "icon",
      size: "small",
      icon: /*#__PURE__*/React.createElement(VisibleLockOutlined, null)
    }),
    ...restProps
  } = props;
  const [visible, setVisible] = useState(false);
  const cssScope = useCssScope('input-password');
  const prefixCls = cssScope();
  const onVisibleChange = () => {
    const {
      disabled
    } = props;
    if (disabled) {
      return;
    }
    setVisible(!visible);
  };
  const preventDefault = e => e.preventDefault();
  const suffixIcon = visibilityToggle && (() => {
    const iconTrigger = ActionMap[action];
    const icon = iconRender(visible);
    return /*#__PURE__*/React.cloneElement(/*#__PURE__*/React.isValidElement(icon) ? icon : /*#__PURE__*/React.createElement("span", null, icon), {
      [iconTrigger]: onVisibleChange,
      key: 'passwordIcon',
      onMouseDown: e => preventDefault(e),
      onMouseUp: e => preventDefault(e)
    });
  })();
  const omittedProps = {
    ...omit(restProps, ['suffix']),
    type: visible ? 'text' : 'password',
    prefixCls,
    suffix: suffixIcon
  };
  return /*#__PURE__*/React.createElement(Input, _extends({
    ref: ref
  }, omittedProps, {
    size: size
  }));
});
