// @universe-design/react/es/input/Search.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React from 'react';
import clsx from 'clsx';
import { Input } from './Input';
import { useComponentSize, useInputGroupSize, useCssScope, useLocale } from '../config-provider/hooks';
import SearchOutlineOutlined from '@universe-design/icons-react/SearchOutlineOutlined';
import { getDefaultSizeAbbr } from '../util';
import './style';
const SearchShapesEnum = ['round', 'square'];

/**
 * 用于搜索输入框 `Input.Search`
 * @public
 */

export const Search = /*#__PURE__*/React.forwardRef((props, ref) => {
  const {
    global: {
      search
    }
  } = useLocale();
  const {
    shape = 'round',
    size: customerSize,
    placeholder = search,
    className,
    ...restProps
  } = props;
  const cssScope = useCssScope('search');
  const configSize = useComponentSize();
  const inputGroupSize = useInputGroupSize();
  const size = customerSize || inputGroupSize || configSize;
  const usedSize = getDefaultSizeAbbr(size);
  const cls = clsx(cssScope(), {
    [cssScope(`--size-${usedSize}`)]: usedSize,
    [cssScope(`--shape-${shape}`)]: shape
  }, className);
  return /*#__PURE__*/React.createElement(Input, _extends({
    ref: ref,
    className: cls,
    size: size,
    placeholder: placeholder,
    prefix: /*#__PURE__*/React.createElement(SearchOutlineOutlined, null)
  }, restProps));
});
if (process.env.NODE_ENV !== "production") {
  Search.displayName = 'Search';
}
