// @universe-design/react/es/input/TextArea.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React, { useCallback, useImperativeHandle, useState } from 'react';
import clsx from 'clsx';
import stylex from '@universe-design/stylex';
import { cssTypographyBody0 } from '../stylex';
import EmptyFilled from '@universe-design/icons-react/EmptyFilled';
import { useIsomorphicLayoutEffect as useLayoutEffect } from '../hooks';
import { useResizeObserver } from '@universe-design/react-utilities';
import { useClassPrefix, useCssScope } from '../config-provider/hooks';
import { isFunction, isNonNullable, KeyCode } from '../util';
import CounterComp from './components/Counter';
import { useInput, useFocus, calculateTextAreaHeight } from '@universe-design/react-input';
import './style';
import { resolveOnChange } from './utils';

/**
 * @public
 */

export const TextArea = /*#__PURE__*/React.forwardRef((props, ref) => {
  const {
    allowClear,
    autoSize,
    showCount,
    maxLength: maxLengthProp,
    value: valueProp,
    defaultValue,
    style,
    error,
    className,
    onKeyDown,
    onPressEnter,
    onChange,
    onClear,
    suffix,
    truncateExcessWords = false,
    counter: counterProp,
    slicer,
    onCompositionEnd: onCompositionEndProp,
    onCompositionStart: onCompositionStartProp,
    autoFocus,
    resize,
    ...resetProps
  } = props;
  const rootPrefixClass = useClassPrefix();
  const cssScope = useCssScope('textarea');
  const [textareaResized, setTextareaResized] = useState(false);
  const textareaRef = useResizeObserver(entry => {
    /**
     * 监控 textarea 的 style 属性上是否有 height 属性
     * textarea 只有在拖拽改变大小之后才会有 style height 属性
     * 此时，更改父容器为 height: auto 以自适应高度
     */
    if (entry.target.style.height) {
      setTextareaResized(true);
    }
  }, {
    disabled: Boolean(autoSize)
  });
  const [textAreaStyle, setTextAreaStyle] = useState({});
  const [focused, setFocused] = useState(false);
  const handleInputChange = useCallback((ref, event, value) => resolveOnChange(ref, event, onChange, value), [onChange]);
  const {
    nativeMaxLength,
    value,
    setValue,
    count,
    onCompositionEnd,
    onCompositionStart,
    internalOnChange,
    maxLength
  } = useInput({
    maxLength: maxLengthProp,
    showCount,
    truncateExcessWords,
    counter: counterProp,
    slicer,
    defaultValue,
    inputRef: textareaRef,
    onChange: handleInputChange,
    onCompositionStart: onCompositionStartProp,
    onCompositionEnd: onCompositionEndProp,
    focused,
    ...('value' in props ? {
      value: valueProp
    } : {})
  });
  useFocus({
    inputRef: textareaRef,
    onFocusChange: f => setFocused(f),
    autoFocus: Boolean(autoFocus)
  });
  useImperativeHandle(ref, () => textareaRef.current);
  const adjustSize = useCallback(() => {
    if (autoSize && textareaRef.current) {
      const config = {
        minRows: undefined,
        maxRows: undefined
      };
      if (typeof autoSize === 'object') {
        config.minRows = autoSize.minRows;
        config.maxRows = autoSize.maxRows;
      }
      const currentStyle = calculateTextAreaHeight(textareaRef.current, false, config.minRows, config.maxRows);
      if (currentStyle) {
        setTextAreaStyle(currentStyle);
      }
    } else {
      setTextAreaStyle({});
    }
  }, [autoSize, textareaRef]);
  const resizeRef = useResizeObserver(adjustSize); // , {disabled: !autoSize})

  useLayoutEffect(adjustSize, [adjustSize, value]);
  const handleClear = e => {
    var _textareaRef$current;
    /* istanbul ignore else  */
    if (!('value' in props)) {
      setValue('');
    }
    /* istanbul ignore else  */
    if (isFunction(onChange) && textareaRef.current) {
      const eventObj = Object.create(e);
      eventObj.target = textareaRef.current;
      eventObj.currentTarget = textareaRef.current;
      const originalValue = textareaRef.current.value;
      textareaRef.current.value = '';
      onChange(eventObj);
      textareaRef.current.value = originalValue;
    }
    (_textareaRef$current = textareaRef.current) === null || _textareaRef$current === void 0 ? void 0 : _textareaRef$current.focus();
    isFunction(onClear) && onClear();
  };
  const showClearIcon = allowClear && !props.disabled && value;
  /**
   * 用户传入的 error true 优先级很高（可能存在输入值不符合某个 form 规则而不只是 maxLength 的情况）
   * count > maxLength error 的优先级也很高
   * 只要有其一就需要标红
   */
  const isError = error || isNonNullable(maxLength) && count > maxLength;
  const cls = clsx({
    [cssScope()]: true,
    [cssScope('--error')]: isError,
    [cssScope('--show-count')]: showCount,
    [cssScope('--show-clear')]: showClearIcon,
    [cssScope('--disabled')]: props.disabled
  }, className);
  return /*#__PURE__*/React.createElement("div", {
    className: cls,
    style: {
      ...style,
      ...(textareaResized ? {
        height: 'auto'
      } : {})
    },
    ref: resizeRef
  }, (() => {
    if (showClearIcon) {
      return /*#__PURE__*/React.createElement("span", {
        className: cssScope('__clear-icon'),
        onClick: handleClear
      }, /*#__PURE__*/React.createElement(EmptyFilled, null));
    }
    return null;
  })(), /*#__PURE__*/React.createElement("textarea", _extends({}, resetProps, {
    style: {
      ...textAreaStyle,
      resize
    },
    ref: textareaRef,
    className: clsx(`${rootPrefixClass}-scrollbar`, stylex(cssTypographyBody0)),
    onChange: internalOnChange,
    onKeyDown: e => {
      /* istanbul ignore else  */
      if (e.keyCode === KeyCode.ENTER) {
        isFunction(onPressEnter) && onPressEnter(e);
      }
      isFunction(onKeyDown) && onKeyDown(e);
    },
    onCompositionEnd: onCompositionEnd,
    onCompositionStart: onCompositionStart,
    value: value,
    maxLength: nativeMaxLength
  })), (() => /*#__PURE__*/React.createElement("span", {
    className: cssScope('-suffix')
  }, suffix && /*#__PURE__*/React.createElement("span", {
    className: cssScope('-suffix-slot')
  }, suffix), /*#__PURE__*/React.createElement(CounterComp, {
    showCount: Boolean(showCount),
    count: count,
    disabled: props.disabled,
    maxLength: maxLengthProp
  })))());
});
