// @universe-design/react/src/stylex/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0GlobalOverrides from '../demo/0.global-overrides';
import C1Typography from '../demo/1.typography';
if (!(C0GlobalOverrides as Story).storyName) {
  (C0GlobalOverrides as Story).storyName = '0. 全局覆盖四级标题字重';
  (C0GlobalOverrides as Story).parameters = {
    docs: {
      description: {
        story:
          'windows 笔记本一般默认命中的字体只有 300 400 700 三种字重，参看 [font-weight 问题解决方案](https://bytedance.feishu.cn/wiki/wikcnUKHN9SnniqQufge5vFSJId#doxcncGem6u40OmuwwDBJwdU)。\n\n例如，由于组件库使用 500 作为四级标题的默认值，在 windows 上会被回退到 400 字重，这会导致 bold 样式丢失。使用 `.ud-typography-title-4` 样式组标注类名可以很方便做到全局覆盖。\n\n',
      },
    },
  };
}
if (!(C1Typography as Story).storyName) {
  (C1Typography as Story).storyName = '1. 版面设计 Typography';
  (C1Typography as Story).parameters = {
    docs: {
      description: {
        story:
          '版面设计相关的「样式组」共有 11 个，以下是他们的类名和默认样式。\n\n\n<style>\n.site-code {\n  display: none !important;\n}\n.site-demo {\n  border: none;\n}\n.site-demo__content {\n  padding: 0;\n}\n</style>\n',
      },
    },
  };
}

export default {
  title: 'Components/Stylex/官网示例',
  tags: ['autodocs'],
};

export { C0GlobalOverrides, C1Typography };
