// @universe-design/react/lib/shared/components/picker/Context.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getPickerContext = exports.PickerContext = void 0;
exports.usePickerContext = usePickerContext;
exports.usePickerContextSelector = void 0;
var _useContextSelector = require("use-context-selector");
var _util = require("../../../util");
const PickerContext = exports.PickerContext = (0, _useContextSelector.createContext)({
  mode: 'date',
  setMode: () => process.env.NODE_ENV !== "production" ? (0, _util.warning)(false, (0, _util.devWarning)('DatePicker', 'You are using the default setMode from context')) : void 0,
  setValue: () => process.env.NODE_ENV !== "production" ? (0, _util.warning)(false, (0, _util.devWarning)('DatePicker', 'You are using the default setValue from context')) : void 0,
  setOpened: () => process.env.NODE_ENV !== "production" ? (0, _util.warning)(false, (0, _util.devWarning)('DatePicker', 'You are using the default setOpened from context')) : void 0,
  showToday: true
});
const getPickerContext = () => PickerContext;
exports.getPickerContext = getPickerContext;
function usePickerContext() {
  return (0, _useContextSelector.useContext)(PickerContext);
}
const usePickerContextSelector = selector => (0, _useContextSelector.useContextSelector)(PickerContext, selector);
exports.usePickerContextSelector = usePickerContextSelector;
