// @universe-design/react/lib/shared/components/picker/Picker.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _noop = _interopRequireDefault(require("lodash/noop"));
var _clsx = _interopRequireDefault(require("clsx"));
var _TimeOutlined = _interopRequireDefault(require("@universe-design/icons-react/TimeOutlined"));
var _useControlled = _interopRequireDefault(require("../../../hooks/useControlled"));
var _util = require("../../../util");
var _ui = require("./utils/ui");
var _useSyncState = _interopRequireDefault(require("./hooks/useSyncState"));
var _composeRefs = _interopRequireDefault(require("./../../../util/composeRefs"));
var _date = require("./utils/date");
var _Context = require("./Context");
var _inputWrapper = _interopRequireDefault(require("./Comp/input-wrapper"));
var _hooks = require("../../../config-provider/hooks");
var _useClickOutsideEffect = _interopRequireDefault(require("./hooks/useClickOutsideEffect"));
var _PickerTrigger = _interopRequireDefault(require("./Comp/PickerTrigger"));
var _PickerPanel = _interopRequireDefault(require("./Comp/PickerPanel"));
var _Input = _interopRequireDefault(require("../Input"));
require("./style");
// Wrap with class component to enable pass generic with instance method
function InnerPicker(props) {
  const timeOperationRef = _react.default.useRef({});
  const {
    locale: localeName,
    components: {
      picker: locale
    }
  } = (0, _hooks.useLocale)();
  const {
    generateConfig,
    triggerElement,
    style,
    className,
    onKeyDown = _noop.default,
    onMouseDown = _noop.default,
    onMouseUp = _noop.default,
    onClick = _noop.default,
    onClear,
    dropdownClassName,
    popupStyle,
    getPopupContainer,
    suffixIcon = /*#__PURE__*/_react.default.createElement(_TimeOutlined.default, null),
    allowClear,
    autoFocus,
    disabled,
    size = 'medium',
    parse,
    picker,
    mode = picker,
    placeholder = locale.timePlaceholder,
    inputReadOnly,
    format = 'HH:mm',
    error,
    onPanelChange
  } = props;
  function dateToText(date) {
    return (0, _ui.formatValue)(generateConfig, date, {
      format,
      mode: picker,
      locale,
      localeName
    });
  }
  const defaultSize = (0, _hooks.useComponentSize)();
  const inputGroupSize = (0, _hooks.useInputGroupSize)();
  const cmpSize = (0, _util.getDefaultSizeAbbr)(size || inputGroupSize || defaultSize);
  const [open, setOpened] = (0, _useControlled.default)(props, {
    defaultValuePropName: '',
    valuePropName: 'open',
    onChangePropName: 'onOpenChange',
    defaultValue: _util.__SCREENSHOT__
  });
  const [focused, setFocused] = (0, _react.useState)(false);

  // eslint-disable-next-line react/destructuring-assignment
  const [value, setValue] = (0, _useControlled.default)(props);
  const [localMode, setLocalMode] = (0, _useSyncState.default)(mode);

  /**
   * 当前视图可见的有效date
   */
  const [viewValue, setViewValue] = (0, _useSyncState.default)(value);
  const validText = dateToText(viewValue);

  /**
   * 输入框的值
   */
  const [text, setText] = (0, _useSyncState.default)(validText);

  /**
   * hover相关的值
   */
  const [hoverDate, setHoverDate] = (0, _react.useState)();
  const hoverText = dateToText(hoverDate);
  const panelDivRef = (0, _react.useRef)(null);
  const inputDivRef = (0, _react.useRef)(null);
  const inputRef = (0, _react.useRef)(null);
  const triggerElementRef = (0, _react.useRef)(null);
  (0, _useClickOutsideEffect.default)((0, _react.useMemo)(() => [panelDivRef, inputDivRef, triggerElementRef], []), open, (0, _react.useCallback)(() => setOpened(false), [setOpened]));
  const cssScope = (0, _hooks.useCssScope)('picker');
  const prefixCls = cssScope();
  const rootCls = (0, _clsx.default)(cssScope(), {
    [cssScope(`--size-${cmpSize}`)]: cmpSize,
    [cssScope('--error')]: error,
    [cssScope('--focus')]: focused
  });
  function confirm(newValue) {
    if (!(0, _date.isEqual)(generateConfig, value, newValue)) {
      setValue(newValue);
    }

    // 暂时同步value值，不然受控模式下，可能出现问题
    setViewValue(value);
    setText(dateToText(value));
    setHoverDate(undefined);
    setOpened(false);
  }
  function handleKeyDown(e) {
    onKeyDown(e);
    if (e.key !== 'Tab' && !open) {
      setOpened(true);
      return;
    }
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    e.keyCode === 13 && confirm(viewValue);
  }
  const handleInputChange = e => {
    setHoverDate(undefined);
    const editText = e.target.value;
    setText(editText);
    const date = (0, _ui.parseValue)(generateConfig, editText, {
      format,
      parse,
      mode,
      locale,
      localeName
    });
    if (date && generateConfig.isValidate(date)
    // && !(typeof disabledDate === 'function' && disabledDate(date))
    ) {
      setViewValue(date);
    }
  };
  const handleClear = () => {
    setValue(undefined);
    // 暂时同步value值，不然受控模式下，可能出现问题
    setViewValue(value);
    setOpened(false);
    onClear && onClear();
  };
  const handleMouseUp = e => {
    if (inputRef.current) {
      inputRef.current.focus();
    }
    setOpened(true);
    onMouseUp(e);
  };
  const handleFocus = e => {
    setFocused(true);
  };
  const handleBlur = e => {
    setFocused(false);
    confirm(viewValue);
  };
  let triggerElementNode;
  if ((0, _util.isFunction)(triggerElement)) {
    triggerElementNode = triggerElement({
      text,
      hoverText,
      open,
      value
    });
  } else {
    triggerElementNode = triggerElement;
  }
  const popupElement = /*#__PURE__*/_react.default.createElement("div", {
    onMouseDown: e => e.preventDefault(),
    ref: panelDivRef
  }, /*#__PURE__*/_react.default.createElement(_Context.PickerContext.Provider, {
    value: {
      mode: localMode,
      setMode: setLocalMode,
      value,
      setValue,
      setOpened
    }
  }, /*#__PURE__*/_react.default.createElement(_PickerPanel.default, (0, _extends2.default)({}, props, {
    prefixCls: prefixCls,
    locale: locale,
    value: viewValue,
    operationRef: timeOperationRef,
    onPanelChange: date => {
      setViewValue(date);
      setText((0, _ui.formatValue)(generateConfig, date, {
        format,
        mode,
        locale,
        localeName
      }));
      onPanelChange && onPanelChange(date);
    }
  }))));
  return /*#__PURE__*/_react.default.createElement(_PickerTrigger.default, {
    open: open,
    disabled: disabled,
    setOpened: setOpened,
    dropdownClassName: dropdownClassName,
    popupStyle: popupStyle,
    popupElement: popupElement,
    getPopupContainer: getPopupContainer
  }, (() => {
    if (triggerElementNode) {
      // 如果用户传递了非 ReactElement 这里会直接报错，导致无法正常渲染
      return /*#__PURE__*/_react.default.cloneElement(triggerElementNode, {
        ref: (0, _composeRefs.default)(triggerElementRef,
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        triggerElementNode.ref)
      });
    }
    return /*#__PURE__*/_react.default.createElement("div", {
      style: style,
      ref: inputDivRef,
      className: rootCls
    }, /*#__PURE__*/_react.default.createElement(_inputWrapper.default, {
      className: (0, _clsx.default)(className, `${prefixCls}-dateInput`),
      showClear: Boolean(viewValue),
      onClear: handleClear,
      prefixCls: prefixCls,
      onMouseUp: handleMouseUp,
      onMouseDown: onMouseDown,
      onClick: onClick,
      focused: focused,
      suffixIcon: suffixIcon,
      disabled: disabled,
      allowClear: allowClear
    }, /*#__PURE__*/_react.default.createElement(_Input.default, {
      size: 12,
      className: (0, _clsx.default)({
        [`${prefixCls}-input`]: true,
        [`${prefixCls}-input-placeholder`]: Boolean(hoverText)
      }),
      placeholder: placeholder,
      disabled: disabled,
      autoFocus: autoFocus,
      ref: inputRef,
      onKeyDown: handleKeyDown,
      onFocus: handleFocus,
      onBlur: handleBlur,
      onChange: handleInputChange,
      readOnly: inputReadOnly || (0, _util.isFunction)(format) && !(0, _util.isFunction)(parse),
      value: hoverText || text
    })));
  })());
}
class Picker extends _react.default.Component {
  constructor() {
    super(...arguments);
    (0, _defineProperty2.default)(this, "pickerRef", /*#__PURE__*/_react.default.createRef());
    (0, _defineProperty2.default)(this, "focus", () => {
      if (this.pickerRef.current) {
        this.pickerRef.current.focus();
      }
    });
    (0, _defineProperty2.default)(this, "blur", () => {
      if (this.pickerRef.current) {
        this.pickerRef.current.blur();
      }
    });
  }
  render() {
    return /*#__PURE__*/_react.default.createElement(InnerPicker, (0, _extends2.default)({}, this.props, {
      pickerRef: this.pickerRef
    }));
  }
}
var _default = exports.default = Picker;
