// @universe-design/react/src/shared/components/highlighter/index.tsx
import React from 'react';
import type { FindAllArgs } from 'highlight-words-core';
import { findAll } from 'highlight-words-core';
import { createContext, useContext } from 'use-context-selector';

const HightLightContext = createContext<{
  searchWords?: string[];
  disabled: boolean;
}>({
  disabled: false,
});

interface TextRenderer {
  (text: string): React.ReactNode;
}

const defaultTextRenderer = (text: string) => <span>{text}</span>;

const defaultMarkRenderer = (text: string) => <mark>{text}</mark>;

export const Highlighter: React.FC<
  FindAllArgs & {
    markRenderer?: TextRenderer;
    textRenderer?: TextRenderer;
  }
> = function ({
  autoEscape = true,
  caseSensitive = false,
  findChunks,
  sanitize,
  searchWords,
  textToHighlight,
  markRenderer = defaultMarkRenderer,
  textRenderer = defaultTextRenderer,
}) {
  const chunks = findAll({
    autoEscape,
    caseSensitive,
    findChunks,
    sanitize,
    searchWords,
    textToHighlight,
  });

  return (
    <span>
      {chunks.map((chunk, index) => {
        const text = textToHighlight.substr(chunk.start, chunk.end - chunk.start);
        if (chunk.highlight) {
          const markNode = markRenderer(text);
          return React.isValidElement(markNode)
            ? React.cloneElement(markNode, { key: index })
            : React.cloneElement(defaultMarkRenderer(text), { key: index });
        }
        const textNode = textRenderer(text);
        return React.isValidElement(textNode)
          ? React.cloneElement(textNode, { key: index })
          : React.cloneElement(defaultTextRenderer(text), { key: index });
      })}
    </span>
  );
};

export const HighlightProvider = HightLightContext.Provider;
if (__DEV__) {
  HighlightProvider.displayName = 'HighlightProvider';
}

export const HighlightConsumer: React.FC<
  React.PropsWithChildren<{
    markRenderer?: TextRenderer;
  }>
> = ({ markRenderer, children }) => {
  const { searchWords, disabled } = useContext(HightLightContext);
  if (typeof children === 'string' && searchWords && searchWords.length && !disabled) {
    return <Highlighter textToHighlight={children} searchWords={searchWords} markRenderer={markRenderer} />;
  }
  return <>{children}</>;
};
if (__DEV__) {
  HighlightConsumer.displayName = 'HighlightConsumer';
}
