// @universe-design/react/src/overflow/hooks.ts
/* eslint-disable max-depth */
/**
The MIT License (MIT)

Copyright (c) 2019-present afc163

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
/**
 * Fork from https://github.com/react-component/overflow/blob/c2ece6ee6e64fcac76c45c3c62a3cdb822159167/src/Overflow.tsx
 * Thanks for https://zhuanlan.zhihu.com/p/340673293
 */
import * as React from 'react';
import { useImmerReducer } from 'use-immer';
import { createContext, useContext, useContextSelector } from 'use-context-selector';
import {
  getPositionAndSize,
  useIsomorphicLayoutEffect as useLayoutEffect,
  // useLogger
} from '../hooks';
import { useResizeObserver } from '@universe-design/react-utilities';

// enableMapSet();

export interface OverflowContextStateType {
  // rest 之前有几个 items
  countBeforeRest: number;
  // rest 之后有几个 items
  countAfterRest: number;
  // rest 宽度
  restWidth: number;
  // prev rest 宽度
  prevRestWidth: number | null;
  // 后缀的起始位置
  suffixFixedStart: number | null;
  // 后缀宽度
  suffixWidth: number;
  // items 宽度
  itemWidths: Record<React.Key, number>;
  // rest 的位置 index
  restPosition: number;
  // 容器宽度
  containerWidth: number;
  // 所有的 keys
  keys: React.Key[];
}

interface OverflowMergedKeys {
  beforeRest: React.Key[];
  afterRest: React.Key[];
}

type OverflowActionType =
  | { type: 'SET_ITEM_WIDTH'; payload: { key: React.Key; width: number } }
  | { type: 'SET_REST_WIDTH'; payload: number }
  | { type: 'SET_SUFFIX_WIDTH'; payload: number }
  | { type: 'SET_CONTAINER_WIDTH'; payload: number }
  | { type: 'SET_KEYS'; payload: React.Key[] };

interface OverflowContextType extends OverflowContextStateType {
  // 所有的 keys
  keys: React.Key[];
  alwaysObserveResize: boolean;
  suffixWrap: boolean;
  dispatch: React.Dispatch<OverflowActionType>;
  omittedKeys: React.Key[];
  isResponsive: boolean;
}

const DEFAULT_STATE: OverflowContextStateType = {
  countBeforeRest: 0,
  countAfterRest: 0,
  restWidth: 0,
  prevRestWidth: null,
  suffixFixedStart: null,
  suffixWidth: 0,
  itemWidths: {},
  restPosition: -1,
  containerWidth: 0,
  keys: [],
};

export const OverflowContext = createContext<OverflowContextType>({
  ...DEFAULT_STATE,
  isResponsive: false,
  alwaysObserveResize: false,
  suffixWrap: true,
  omittedKeys: [],
  dispatch: () => void 0,
});

if (__DEV__) {
  OverflowContext.displayName = 'OverflowContext';
}

const getMergedKeys = ({
  keys,
  allCount,
  restPosition,
}: {
  /**
   * 传入的所有 keys
   */
  keys: React.Key[];
  /**
   * 最大能展开计算的数量
   */
  allCount: number;
  /**
   * rest 节点的位置
   */
  restPosition: number;
}): OverflowMergedKeys => {
  const beforeRest = keys.slice(0, Math.min(restPosition, allCount));
  let afterRest: React.Key[] = [];
  if (beforeRest.length < allCount) {
    afterRest = keys.slice(keys.length - allCount + beforeRest.length, keys.length);
  }
  return {
    beforeRest,
    afterRest,
  };
};

const getWidth = (target?: Element | null, entry?: ResizeObserverEntry | undefined) =>
  target ? getPositionAndSize(target, entry).width : 0;

/**
 * ## 1. 初始渲染节点数量
 *
 * 一开始不能渲染所有的节点，如果项目太多渲染压力会比较大，所以一开始就需要大致计算出一个合理的数值。这里我们采用如下的算法计算初始值
 *
 * 1. 获取用户提供项目的可能的最小宽度，如果未提供，使用默认值 10
 * 2. 取 keys 长度和 (containerWidth / 10) 两者的较小值作为渲染的项目数量
 *
 * ## 2. 有限次计算得到稳态
 *
 * 如果一开始放开的节点足够多，那么最少两次 rerender 就可以得到稳态，如果一开始放开的节点不够多，就可能需要不断得多次渲染来得到稳态（这里可以通过已渲染的元素宽度来做优化）。
 *
 * ### 词汇表
 *
 * ### 流程
 *
 * 尽可能去展开，所以所有的判断结束都要在超出时判断。
 *
 * 1. 如果没有 keys，就直接显示 suffix，结束。
 * 2. 计算 totalWidth 初始值 totalWidth = suffixWidth。
 * 2. 执行宽度累加子流程：
 *    1. 判断是否已经有 itemWidth，如果没有则放宽显示 keys 至当前 index，结束。
 *    2. 计算当前累加宽度值，totalWidth = totalWidth + itemWidth_i。
 *    3. 如果 totalWidth + restWidth（如果 i = maxIndex，restWidth 按 0 计算） > containerWidth，说明当前 i 不可用，最终显示到 i-1，suffix 位置为 sum(itemWidth_i-1) + restWidth，结束（这里不一定真的结束，因为 restWidth 有可能变，导致再次执行该过程）
 *    4. 如果不符合 3，但是 i 是 lastIndex，那么就可以显示到 i，suffix 位置为 sum(itemWidth_i)
 */
export function useOverflow<C extends HTMLElement>({
  keys,
  maxCount,
  hasSuffix,
  restPosition = -1,
  suffixWrap,
}: {
  keys: React.Key[];
  maxCount?: number | 'responsive';
  hasSuffix: boolean;
  restPosition?: number;
  suffixWrap: boolean;
}) {
  const isResponsive = Boolean(keys.length) && maxCount === 'responsive';

  const mergedRestPosition = restPosition < 0 ? restPosition + keys.length + 1 : restPosition;

  const getCountsFromDisplayCount = (_displayCount: number) => {
    if (_displayCount > mergedRestPosition) {
      return {
        countBeforeRest: mergedRestPosition,
        countAfterRest: _displayCount - mergedRestPosition,
      };
    }
    return {
      countBeforeRest: _displayCount,
      countAfterRest: 0,
    };
  };

  const [state, dispatch] = useImmerReducer<OverflowContextStateType, OverflowActionType>((draft, action) => {
    // console.log('useOverflow', action);

    const setDisplayCount = () => {
      // 核心方法
      // pos = 1
      // 0 1 2 3 4 5
      // 0 ... 2 3 4 5
      // 0 ... 3 4 5
      // 0 ... 4 5
      // 0 ... 5
      // ... 5
      // ...

      // pos = -1
      // 0 1 2 3 4 5
      // 0 1 2 3 4 ...
      // 0 1 2 3 ...
      // 0 1 2 ...
      // 0 1 ...
      // 0 ...
      // ...

      // const displayCount = countBeforeRest + countAfterRest;

      // console.log({ containerWidth, restWidth });

      // Always use the max width to avoid blink
      const mergedRestWidth = Math.max(draft.prevRestWidth ?? draft.restWidth, draft.restWidth);
      draft.prevRestWidth = draft.restWidth;

      let totalWidth = draft.suffixWidth;

      const mergedKeys = getMergedKeys({
        keys,
        allCount: Math.min(keys.length, Math.ceil(draft.containerWidth / 10)),
        restPosition: mergedRestPosition,
      });
      const _mergedKeys = [...mergedKeys.beforeRest, ...[...mergedKeys.afterRest].reverse()];
      const len = _mergedKeys.length;
      const lastIndex = len - 1;

      interface IterData {
        i: number;
        right: number;
        suffixFixedStart: number | null;
      }

      /**
       * 最佳显示 iteration 的数据
       */
      const allCollapsedWidth = totalWidth + mergedRestWidth;
      let bestIter:
        | {
            nowrap: IterData;
            wrap: IterData | null;
          }
        | {
            nowrap: IterData | null;
            wrap: IterData;
          };
      const defaultIter = {
        i: -1,
        right: allCollapsedWidth,
        suffixFixedStart: mergedRestWidth,
      };
      if (allCollapsedWidth - draft.containerWidth > 0) {
        bestIter = {
          wrap: defaultIter,
          nowrap: null,
        };
      } else {
        bestIter = {
          wrap: null,
          nowrap: defaultIter,
        };
      }

      const updateBestIter = (iterData: IterData) => {
        // nowrap 更新
        if (iterData.right - draft.containerWidth <= 0) {
          if (!bestIter.nowrap || iterData.i > bestIter.nowrap.i) {
            bestIter.nowrap = iterData;
          }
          return;
        }
        // wrap 更新
        if (iterData.right - draft.containerWidth > 0) {
          if (
            iterData.right - draft.suffixWidth - draft.containerWidth <= 0 &&
            (!bestIter.wrap || iterData.i > bestIter.wrap.i)
          ) {
            bestIter.wrap = iterData;
          }
          if (!bestIter.wrap || iterData.right <= bestIter.wrap.right) {
            bestIter.wrap = iterData;
          }
          return;
        }
      };

      for (const [i, key] of _mergedKeys.entries()) {
        const itemWidth = draft.itemWidths[key];

        if (itemWidth === undefined) {
          const { countAfterRest, countBeforeRest } = getCountsFromDisplayCount(i);
          draft.countAfterRest = countAfterRest;
          draft.countBeforeRest = countBeforeRest;
          break;
        }

        totalWidth += itemWidth;
        const right = totalWidth + (i === lastIndex ? 0 : mergedRestWidth);

        // console.table({
        //   i,
        //   containerWidth,
        //   totalWidth,
        //   itemWidth,
        //   mergedRestWidth,
        //   right,
        //   mergedRestPosition,
        // });

        updateBestIter({
          i,
          right,
          suffixFixedStart: right - draft.suffixWidth,
        });
      }

      // eslint-disable-next-line one-var
      let _bestIter: IterData;
      if (suffixWrap) {
        _bestIter = bestIter.nowrap ?? bestIter.wrap ?? defaultIter;
      } else {
        _bestIter = bestIter.nowrap ?? defaultIter;
      }

      const { countAfterRest, countBeforeRest } = getCountsFromDisplayCount(_bestIter.i + 1);
      draft.countAfterRest = countAfterRest;
      draft.countBeforeRest = countBeforeRest;

      if (hasSuffix && _bestIter.right > draft.containerWidth) {
        draft.suffixFixedStart = null;
      } else {
        draft.suffixFixedStart = _bestIter.suffixFixedStart;
      }

      return;
    };

    switch (action.type) {
      case 'SET_ITEM_WIDTH':
        if (action.payload.width !== undefined) {
          draft.itemWidths[action.payload.key] = action.payload.width;
        } else {
          delete draft.itemWidths[action.payload.key];
        }
        setDisplayCount();
        return;
      case 'SET_REST_WIDTH':
        draft.restWidth = action.payload;
        setDisplayCount();
        return;
      case 'SET_SUFFIX_WIDTH':
        draft.suffixWidth = action.payload;
        setDisplayCount();
        return;
      case 'SET_CONTAINER_WIDTH':
        draft.containerWidth = action.payload;
        setDisplayCount();
        return;
      case 'SET_KEYS':
        draft.keys = action.payload;
        setDisplayCount();
        return;
      default:
        return;
    }
  }, DEFAULT_STATE);

  const ref = useResizeObserver<C>(entry => {
    const width = getWidth(entry.target, entry);
    dispatch({ type: 'SET_CONTAINER_WIDTH', payload: width });
  });

  useLayoutEffect(() => {
    // 每个组件都要在 useLayoutEffect 里面获取一次值，为了让渲染能阻塞在本组件内，不向上扩散，这样用户第一次 mount 就能拿到正确的尺寸
    const width = getWidth(ref.current);
    dispatch({ type: 'SET_CONTAINER_WIDTH', payload: width });
  }, [dispatch, ref]);

  useLayoutEffect(() => {
    // 这里在每次 keys 变化时，都重新计算一次，确保当 items 被删除时，缓存值会被清理，触发重新计算逻辑，从而正确更新 suffixFixedStart 的位置
    dispatch({ type: 'SET_KEYS', payload: keys });
  }, [dispatch, isResponsive, keys]);

  const mergedKeys = React.useMemo<OverflowMergedKeys>(() => {
    const keys = state.keys;
    if (isResponsive) {
      return getMergedKeys({
        keys,
        allCount: Math.min(keys.length, Math.ceil(state.containerWidth / 10)),
        restPosition: mergedRestPosition,
      });
    }
    return getMergedKeys({
      keys,
      allCount: (() => {
        if (typeof maxCount === 'number') {
          return Math.max(Math.min(maxCount, keys.length), 0);
        }
        return keys.length;
      })(),
      restPosition: mergedRestPosition,
    });
  }, [state.keys, state.containerWidth, isResponsive, mergedRestPosition, maxCount]);

  const omittedKeys = React.useMemo(() => {
    const keys = state.keys;
    if (isResponsive) {
      return keys.slice(state.countBeforeRest, keys.length - state.countAfterRest);
    }
    return keys.slice(mergedKeys.beforeRest.length, keys.length - mergedKeys.afterRest.length);
  }, [
    isResponsive,
    state.keys,
    state.countBeforeRest,
    state.countAfterRest,
    mergedKeys.beforeRest.length,
    mergedKeys.afterRest.length,
  ]);

  return {
    ref,
    ...state,
    mergedKeys,
    omittedKeys,
    dispatch,
    isResponsive,
    restPosition: mergedRestPosition,
    suffixWrap,
  };
}

export function useOverflowItem<C extends HTMLElement>({ key, index }: { key: React.Key; index: number }) {
  const dispatch = useContextSelector(OverflowContext, o => o.dispatch);
  const alwaysObserveResize = useContextSelector(OverflowContext, o => o.alwaysObserveResize);
  const length = useContextSelector(OverflowContext, o => o.keys.length);
  const countBeforeRest = useContextSelector(OverflowContext, o => o.countBeforeRest);
  const countAfterRest = useContextSelector(OverflowContext, o => o.countAfterRest);
  const isResponsive = useContextSelector(OverflowContext, o => o.isResponsive);
  const ref = useResizeObserver<C>(entry => {
    const width = getWidth(entry.target, entry);
    // 注册 item width
    dispatch({
      type: 'SET_ITEM_WIDTH',
      payload: {
        key,
        width,
      },
    });
  });
  useLayoutEffect(() => {
    const width = getWidth(ref.current);
    dispatch({
      type: 'SET_ITEM_WIDTH',
      payload: {
        key,
        width,
      },
    });
  }, [dispatch, key, ref]);
  // useLogger('useOverflowItem', { key, index });
  return {
    ref: isResponsive || alwaysObserveResize ? ref : undefined,
    isVisible: (isResponsive && (index > length - countAfterRest - 1 || index < countBeforeRest)) || !isResponsive,
    isResponsive,
  };
}

export function useOverflowRest<C extends HTMLElement>() {
  const dispatch = useContextSelector(OverflowContext, o => o.dispatch);
  const alwaysObserveResize = useContextSelector(OverflowContext, o => o.alwaysObserveResize);
  const length = useContextSelector(OverflowContext, o => o.keys.length);
  const countBeforeRest = useContextSelector(OverflowContext, o => o.countBeforeRest);
  const countAfterRest = useContextSelector(OverflowContext, o => o.countAfterRest);
  const isResponsive = useContextSelector(OverflowContext, o => o.isResponsive);
  const restPosition = useContextSelector(OverflowContext, o => o.restPosition);
  const ref = useResizeObserver<C>(entry => {
    const width = getWidth(entry.target, entry);
    dispatch({
      type: 'SET_REST_WIDTH',
      payload: width,
    });
  });
  useLayoutEffect(() => {
    const width = getWidth(ref.current);
    dispatch({
      type: 'SET_REST_WIDTH',
      payload: width,
    });
  }, [dispatch, ref]);
  // useLogger('useOverflowRest');
  return {
    ref: isResponsive || alwaysObserveResize ? ref : undefined,
    countBeforeRest,
    countAfterRest,
    isVisible: (isResponsive && countBeforeRest + countAfterRest < length) || !isResponsive,
    isResponsive,
    restPosition,
  };
}

export function useOverflowSuffix<C extends HTMLElement>() {
  const dispatch = useContextSelector(OverflowContext, o => o.dispatch);
  const alwaysObserveResize = useContextSelector(OverflowContext, o => o.alwaysObserveResize);
  const countBeforeRest = useContextSelector(OverflowContext, o => o.countBeforeRest);
  const countAfterRest = useContextSelector(OverflowContext, o => o.countAfterRest);
  const isResponsive = useContextSelector(OverflowContext, o => o.isResponsive);
  const suffixFixedStart = useContextSelector(OverflowContext, o => o.suffixFixedStart);
  const restWidth = useContextSelector(OverflowContext, o => o.restWidth);
  const suffixWrap = useContextSelector(OverflowContext, o => o.suffixWrap);
  const ref = useResizeObserver<C>(entry => {
    const width = getWidth(entry.target, entry);
    dispatch({
      type: 'SET_SUFFIX_WIDTH',
      payload: width,
    });
  });
  useLayoutEffect(() => {
    const width = getWidth(ref.current);
    dispatch({
      type: 'SET_SUFFIX_WIDTH',
      payload: width,
    });
  }, [dispatch, ref]);
  // useLogger('useOverflowSuffix');
  return {
    ref: isResponsive || alwaysObserveResize ? ref : undefined,
    countBeforeRest,
    countAfterRest,
    isResponsive,
    suffixFixedStart,
    restWidth,
    suffixWrap,
  };
}

export const useOverflowContext = () => useContext(OverflowContext);
