// @universe-design/react/src/overflow/index.tsx
import clsx from 'clsx';
import type { CSSProperties, FC, Key, PropsWithChildren } from 'react';
import React from 'react';
import { useCssScope } from '../config-provider/hooks';
import { useOverflowItem, OverflowContext, useOverflow, useOverflowRest, useOverflowSuffix } from './hooks';

import './style';

export interface OverflowProps<K extends Key> {
  style?: CSSProperties;
  getItemStyle?: (key: K) => CSSProperties;
  restStyle?: CSSProperties;
  className?: string;
  keys: K[];
  maxCount?: number | 'responsive';
  itemRenderer?: (key: K) => React.ReactNode;
  restRenderer?: (omittedKeys: K[]) => React.ReactNode;
  suffixRenderer?: (ctx: { omittedKeys: K[] }) => React.ReactNode;
  addonRenderer?: (ctx: { omittedKeys: K[] }) => React.ReactNode;
  suffixClassName?: string;
  restPosition?: number;
  /**
   * 在非 responsive 模式下也进行 resize observe
   * @internal
   */
  alwaysObserveResize?: boolean;
  /**
   * 是否允许 suffix 换行
   * @defaultValue true
   */
  suffixWrap?: boolean;
}

const _OverflowItem = <K extends Key>({
  value,
  index,
  children,
  getItemStyle,
}: PropsWithChildren<{ value: K; index: number; getItemStyle?: (key: K) => CSSProperties }>) => {
  const cssScope = useCssScope('overflow__item');
  const { ref, isVisible } = useOverflowItem<HTMLDivElement>({
    key: value,
    index,
  });
  const itemStyle = getItemStyle?.(value);
  const order = isVisible ? index : Number.MAX_SAFE_INTEGER;
  return (
    <div
      className={cssScope()}
      style={{
        order,
        height: isVisible ? undefined : 0,
        overflowY: isVisible ? undefined : 'hidden',
        pointerEvents: isVisible ? undefined : 'none',
        ...itemStyle,
      }}
    >
      <div ref={ref}>{children}</div>
    </div>
  );
};

if (__DEV__) {
  _OverflowItem.displayName = 'OverflowItem';
}

const OverflowItem = React.memo(_OverflowItem) as typeof _OverflowItem;

const OverflowRest: FC<PropsWithChildren<{ restStyle?: CSSProperties }>> = React.memo(({ children, restStyle }) => {
  const { ref, countBeforeRest, isVisible, isResponsive, restPosition } = useOverflowRest<HTMLDivElement>();
  const cssScope = useCssScope('overflow__rest');
  const order = isVisible ? countBeforeRest - 1 : Number.MAX_SAFE_INTEGER;
  return (
    <div
      ref={ref}
      className={cssScope('')}
      style={{
        // 这里是为了修复在父级 fit-content 时，rest 节点没有展示也会占空间的问题
        position: isVisible ? undefined : 'absolute',
        order: isResponsive ? order : restPosition - 1,
        height: isVisible ? undefined : 0,
        overflowY: isVisible ? undefined : 'hidden',
        pointerEvents: isVisible ? undefined : 'none',
        ...restStyle,
      }}
    >
      {children}
    </div>
  );
});

const OverflowSuffix: FC<
  PropsWithChildren<{
    className?: string;
  }>
> = React.memo(({ className, children }) => {
  const { ref, suffixFixedStart, countAfterRest, countBeforeRest, isResponsive, restWidth, suffixWrap } =
    useOverflowSuffix<HTMLDivElement>();
  const cssScope = useCssScope('overflow__suffix');
  const order = countBeforeRest + countAfterRest - 1;
  return (
    <div
      className={clsx(className, cssScope(''))}
      ref={ref}
      style={{
        order: isResponsive ? order : Number.MAX_SAFE_INTEGER,
        ...(suffixFixedStart !== null && isResponsive
          ? {
              position: 'absolute',
              left: suffixFixedStart,
            }
          : {}),
        maxWidth: suffixWrap ? undefined : `calc(100% - ${restWidth}px)`, // 限制 suffix 不换行时的最大宽度
      }}
    >
      {children}
    </div>
  );
});

const defaultItemRenderer = (key: Key) => (
  <div
    style={{
      border: '1px red solid',
    }}
  >
    {Array(Number(key) + 1)
      .fill(0)
      .map((_, i) => i)
      .join(',')}
  </div>
);

const defaultRestRenderer = (omittedKeys: Key[]) => (
  <div
    style={{
      border: '1px blue solid',
    }}
  >
    {`+${omittedKeys.length}`}
  </div>
);

export const Overflow = <K extends Key = Key>({
  maxCount,
  keys,
  style,
  getItemStyle,
  restStyle,
  className,
  itemRenderer = defaultItemRenderer,
  restRenderer = defaultRestRenderer,
  suffixRenderer,
  suffixClassName,
  restPosition,
  addonRenderer,
  alwaysObserveResize = false,
  suffixWrap = true,
}: OverflowProps<K>) => {
  const cssScope = useCssScope('overflow');
  const { ref, dispatch, isResponsive, ...state } = useOverflow<HTMLDivElement>({
    maxCount,
    keys,
    hasSuffix: Boolean(suffixRenderer),
    restPosition,
    suffixWrap,
  });

  const showRest = isResponsive || (typeof maxCount === 'number' && keys.length > maxCount);
  return (
    <OverflowContext.Provider value={{ dispatch, isResponsive, ...state, alwaysObserveResize }}>
      <div ref={ref} className={clsx(cssScope(), className)} style={style}>
        {[...state.mergedKeys.beforeRest, ...state.mergedKeys.afterRest].map((key, i) => (
          <OverflowItem<K> key={key} value={key as K} index={i} getItemStyle={getItemStyle}>
            {itemRenderer(key as K)}
          </OverflowItem>
        ))}
        {showRest && <OverflowRest restStyle={restStyle}>{restRenderer(state.omittedKeys as K[])}</OverflowRest>}
        {suffixRenderer && (
          <OverflowSuffix className={suffixClassName}>
            {suffixRenderer({ omittedKeys: state.omittedKeys as K[] })}
          </OverflowSuffix>
        )}
        {addonRenderer?.({ omittedKeys: state.omittedKeys as K[] })}
      </div>
    </OverflowContext.Provider>
  );
};
