// @universe-design/react/src/menu/internal-menu/context/MenuContext.tsx
import * as React from 'react';
import type { CSSMotionProps } from '@universe-design/rc-motion';
import useMemo from '@universe-design/rc-util/lib/hooks/useMemo';
import { shallowEqualObjects as shallowEqual } from 'shallow-equal';
import type {
  BuiltinPlacements,
  MenuClickEventHandler,
  RcMenuMode,
  RenderIconType,
  TriggerSubMenuAction,
} from '../interface';

export interface RcMenuContextProps {
  prefixCls: string;
  openKeys: string[];
  rtl?: boolean;

  // Mode
  mode: RcMenuMode;

  // Disabled
  disabled?: boolean;
  // Used for overflow only. Prevent hidden node trigger open
  overflowDisabled?: boolean;

  // Active
  activeKey: string;
  onActive: (key: string) => void;
  onInactive: (key: string) => void;

  // Selection
  selectedKeys: string[];

  // Level
  inlineIndent: number;

  // Motion
  motion?: CSSMotionProps;
  defaultMotions?: Partial<{ [key in RcMenuMode | 'other']: CSSMotionProps }>;

  // Popup
  subMenuOpenDelay: number;
  subMenuCloseDelay: number;
  forceSubMenuRender?: boolean;
  builtinPlacements?: BuiltinPlacements;
  triggerSubMenuAction?: TriggerSubMenuAction;

  // Icon
  itemIcon?: RenderIconType;
  expandIcon?: RenderIconType;

  // Function
  onItemClick: MenuClickEventHandler;
  onOpenChange: (key: string, open: boolean) => void;
  getPopupContainer: (node: HTMLElement) => HTMLElement;
}

// @ts-expect-error: RC 遗留问题
export const RcMenuContext = React.createContext<RcMenuContextProps>(null);

function mergeProps(origin: RcMenuContextProps, target: Partial<RcMenuContextProps>): RcMenuContextProps {
  const clone: RcMenuContextProps = { ...origin };
  Object.entries(target).forEach(([key, value]) => {
    if (value !== undefined) {
      // @ts-expect-error: igonre
      clone[key] = value;
    }
  });

  // (Object.keys(target) as (keyof MenuContextProps)[]).forEach(key => {
  //   const value = target[key];
  //   if (value !== undefined) {
  //     clone[key] = value;
  //   }
  // });

  return clone;
}

export interface InheritableContextProps extends Partial<RcMenuContextProps> {
  children?: React.ReactNode;
  locked?: boolean;
}

export default function InheritableContextProvider({ children, locked, ...restProps }: InheritableContextProps) {
  const context = React.useContext(RcMenuContext);

  const inheritableContext = useMemo(
    () => mergeProps(context!, restProps),
    [context, restProps],
    (prev, next) => !locked && (prev[0] !== next[0] || !shallowEqual(prev[1], next[1])),
  );

  return <RcMenuContext.Provider value={inheritableContext}>{children}</RcMenuContext.Provider>;
}
