// @byted-service/bytedtrace/lib/confProvider/configer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Configer = void 0;
const tslib_1 = require("tslib");
const consts = tslib_1.__importStar(require("./consts"));
const remote = tslib_1.__importStar(require("./remote"));
const assert_1 = require("assert");
const asyncCache_1 = require("./asyncCache");
class Configer {
    constructor(namespace, event, logger) {
        this.event = event;
        this.namespace = namespace;
        this.defaultConfig = Object.assign({}, consts.DEFAULT_CONFIGS[this.namespace]);
        this.cache = new asyncCache_1.AsyncCache({
            fetcher: this.fetch.bind(this),
            isSame: this.isSame.bind(this),
            changeHandler: this.notify.bind(this),
            logger,
        });
        this.fetcher = new remote.Fetcher(this.namespace, new remote.KVStore(), this.defaultConfig);
    }
    async asyncSetDefaultConfig() {
        this.cache.setDefault(this.namespace, this.defaultConfig);
        this.cache.getOrSet(this.namespace, this.defaultConfig);
    }
    async syncSetDefaultConfig() {
        await this.cache.getOrSet(this.namespace, this.defaultConfig);
    }
    // get gets circuit breaker config from async cache.
    async get(force) {
        try {
            const res = await this.cache.get(this.namespace, force);
            return res;
        }
        catch (err) {
            /* istanbul ignore next */
            return this.defaultConfig;
        }
    }
    // dump dumps global tracer config from async cache.
    dump() {
        return this.cache.dump();
    }
    // isSame determines new circuit breaker config is same or not.
    isSame(oldValue, newValue) {
        if (!newValue) {
            return true;
        }
        if (!oldValue) {
            return false;
        }
        try {
            (0, assert_1.deepStrictEqual)(oldValue, newValue);
            return true;
        }
        catch (e) {
            /* istanbul ignore next */
            return false;
        }
    }
    // notify notifies global tracer config change update event.
    notify(newValue) {
        if (newValue) {
            this.event.emit(consts.EVENT_CONFIG_CHANGE, {
                namespace: this.namespace,
                config: newValue,
                timestamp: Date.now(),
            });
        }
    }
    // fetch fetches global tracer config by fetcher.
    async fetch() {
        return await this.fetcher.fetch();
    }
}
exports.Configer = Configer;
