// @universe-design/react/lib/menu/internal-menu/SubMenu/PopupTrigger.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = PopupTrigger;
var React = _interopRequireWildcard(require("react"));
var _rcTrigger = _interopRequireDefault(require("@universe-design/rc-trigger"));
var _clsx = _interopRequireDefault(require("clsx"));
var _raf = _interopRequireDefault(require("@universe-design/rc-util/lib/raf"));
var _MenuContext = require("../context/MenuContext");
var _placements = require("../placements");
var _motionUtil = require("../utils/motionUtil");
const popupPlacementMap = {
  horizontal: 'bottomLeft',
  vertical: 'rightTop',
  'vertical-left': 'rightTop',
  'vertical-right': 'leftTop'
};
function PopupTrigger(_ref) {
  let {
    prefixCls,
    visible,
    children,
    popup,
    popupClassName,
    popupOffset,
    disabled,
    mode,
    onVisibleChange
  } = _ref;
  const {
    getPopupContainer,
    rtl,
    subMenuOpenDelay,
    subMenuCloseDelay,
    builtinPlacements,
    triggerSubMenuAction,
    forceSubMenuRender,
    // Motion
    motion,
    defaultMotions
  } = React.useContext(_MenuContext.RcMenuContext);
  const [innerVisible, setInnerVisible] = React.useState(false);
  const placement = rtl ? {
    ..._placements.rcPlacementsRtl,
    ...builtinPlacements
  } : {
    ..._placements.rcPlacements,
    ...builtinPlacements
  };
  const popupPlacement = popupPlacementMap[mode];
  const targetMotion = (0, _motionUtil.getMotion)(mode, motion, defaultMotions);
  const mergedMotion = {
    ...targetMotion,
    leavedClassName: `${prefixCls}-hidden`,
    removeOnLeave: false,
    motionAppear: true
  };

  // Delay to change visible
  const visibleRef = React.useRef();
  React.useEffect(() => {
    visibleRef.current = (0, _raf.default)(() => {
      setInnerVisible(visible);
    });
    return () => {
      _raf.default.cancel(visibleRef.current);
    };
  }, [visible]);
  return /*#__PURE__*/React.createElement(_rcTrigger.default, {
    prefixCls: prefixCls,
    popupClassName: (0, _clsx.default)(`${prefixCls}-popup`, {
      [`${prefixCls}-rtl`]: rtl
    }, popupClassName),
    stretch: mode === 'horizontal' ? 'minWidth' : undefined,
    getPopupContainer: getPopupContainer,
    builtinPlacements: placement,
    popupPlacement: popupPlacement,
    popupVisible: innerVisible,
    popup: popup,
    popupAlign: popupOffset && {
      offset: popupOffset
    },
    action: disabled ? [] : triggerSubMenuAction && [triggerSubMenuAction],
    mouseEnterDelay: subMenuOpenDelay,
    mouseLeaveDelay: subMenuCloseDelay,
    onPopupVisibleChange: onVisibleChange,
    forceRender: forceSubMenuRender,
    popupMotion: mergedMotion
  }, children);
}
