// @universe-design/react/src/select/stories/homeless.stories.tsx
/* eslint-disable react/no-children-prop */
/* eslint-disable max-lines-per-function */
/** @jsxImportSource @emotion/react */
import type { RefObject } from 'react';
import React, { useState } from 'react';
// also exported from '@storybook/react' if you can deal with breaking changes in 6.1
import type { Meta } from '@storybook/react';

import orderBy from 'lodash/orderBy';
import type { SelectOptionGroupConfig, SelectProps, SelectRef } from '../..';
import { Avatar, Button, Checkbox, Col, Input, Row, Select } from '../..';
import { Highlighter } from '../../shared';

import All from '../demo/all';

export default {
  title: 'Components/Select/临时示例',
  component: Select,
} satisfies Meta;

const { Option } = Select;

const options = [
  {
    value: 'Jack',
  },
  {
    value: 'Lucy',
  },
  {
    value: 'Tom',
  },
];

export const SelectBase = () => {
  const [showEmpty, setShowEmpty] = React.useState(false);
  const [list, setList] = React.useState<
    {
      name: string;
      alpha2Code: string;
      region: string;
      subregion: string;
      population: number;
      area: number;
    }[]
  >([]);
  const processedList = React.useMemo(() => (showEmpty ? [] : list), [showEmpty, list]);
  const options = React.useMemo(() => {
    const ret = [] as NonNullable<SelectProps<string>['options']>;
    const groupMap = new Map<string, SelectOptionGroupConfig<string>>();
    processedList.forEach(item => {
      const firstChar = item.name[0];
      let group = groupMap.get(firstChar);
      if (!group) {
        group = {
          key: firstChar,
          label: firstChar,
          options: [],
        };
        ret.push(group);
        groupMap.set(firstChar, group);
      }
      group.options.push({
        value: item.name,
        disabled: Math.random() > 0.5,
      });
    });
    return orderBy(ret, o => ('options' in o ? o.key : o.value));
  }, [processedList]);
  React.useEffect(() => {
    fetch('https://restcountries.eu/rest/v2/all')
      .then(resp => resp.json())
      .then(setList);
  }, []);
  const selectRef = React.useRef<SelectRef<string>>(null);
  return (
    <div>
      {/* <Global
        styles={{
          body: {
            backgroundColor: '#F5F6F7',
          },
        }}
      /> */}
      <Checkbox onClick={() => (showEmpty ? setShowEmpty(false) : setShowEmpty(true))} />
      <h4>Size</h4>
      <All />
      <h4>Status</h4>
      <Row gutter={16}>
        <Col>
          <h6>empty</h6>
          <Select showSearch={true} style={{ width: '200px' }} placeholder="Select a country" options={[]} />
        </Col>
        <Col>
          <h6>disabled</h6>
          <Select value="Chile" disabled style={{ width: '200px' }} placeholder="Select a country" options={options} />
        </Col>
        <Col>
          <h6>error</h6>
          <p>
            <button type="button" onClick={() => selectRef.current?.focus()}>
              Focus
            </button>
          </p>
          <Select
            ref={selectRef}
            error
            value="Chile"
            style={{ width: '200px' }}
            placeholder="Select a country"
            options={options}
          />
        </Col>
        <Col>
          <h6>danger disabled</h6>
          <Select
            error
            disabled
            value="United States of America"
            style={{ width: '200px' }}
            placeholder="Select a country"
            options={options}
          />
        </Col>
      </Row>
      <h4>With Search</h4>
      <Row gutter={16}>
        <Col>
          <Select style={{ width: '200px' }} placeholder="Select a country" options={options} />
        </Col>
        <Col>
          <Select bordered="underlined" style={{ width: '200px' }} placeholder="Select a country" options={options} />
        </Col>
      </Row>
      <h4>Dropdown Width</h4>
      <Row gutter={16}>
        <Col>
          <h6>dropdownMatchSelectWidth: false</h6>
          <Select
            dropdownMatchSelectWidth={false}
            style={{ width: '200px' }}
            virtual={false}
            placeholder="Select a country"
            options={options}
            showSearch
          />
        </Col>
        <Col>
          <h6>dropdownMatchSelectWidth: 420</h6>
          <Select
            dropdownMatchSelectWidth={420}
            style={{ width: '200px' }}
            placeholder="Select a country"
            options={options}
            showSearch
          />
        </Col>
      </Row>
      <h4>Selector Adaptive Width</h4>
      <Row>
        <Col>
          <Select style={{ width: 'max-content' }} placeholder="Select a country" options={options} showSearch />
        </Col>
      </Row>
      <h4>Multiple Select</h4>
      <Row gutter={16}>
        <Col>
          <h6>normal</h6>
          {options.length ? (
            <Select
              defaultValue={['China', 'Chile', 'Albania', 'Uzbekistan']}
              style={{ width: '200px' }}
              placeholder="Select a country"
              mode="multiple"
              options={options}
            />
          ) : null}
        </Col>
        <Col>
          <h6>disabled</h6>
          <Select
            value={['China', 'Chile', 'Albania', 'Uzbekistan']}
            disabled
            style={{ width: '200px' }}
            placeholder="Select a country"
            maxTagCount="responsive"
            mode="multiple"
            options={options}
          />
        </Col>
        <Col>
          <h6>no wrap</h6>
          {options.length ? (
            <Select
              allowClear
              defaultValue={['China', 'Chile', 'Albania', 'Uzbekistan']}
              style={{ width: '200px' }}
              placeholder="Select a country"
              maxTagCount="responsive"
              mode="multiple"
              options={options}
            />
          ) : null}
        </Col>
      </Row>
      <h4>Customized Select Option</h4>
      <Row gutter={16}>
        <Col>
          <Select
            style={{ width: '200px' }}
            placeholder="Select a country"
            mode="multiple"
            options={list.map(item => ({
              value: item.name,
              labelRenderer: (inputValue: string) => (
                <div
                  css={{
                    display: 'inline-flex',
                    alignItems: 'flex-start',
                    width: '100%',
                  }}
                >
                  <Avatar css={{ marginRight: 8, flex: 'none' }} children={item.alpha2Code} />
                  <div
                    css={{
                      overflow: 'hidden',
                      textOverflow: 'ellipsis',
                      whiteSpace: 'nowrap',
                    }}
                  >
                    <Highlighter
                      searchWords={[inputValue]}
                      textToHighlight={item.name}
                      markRenderer={text => <span css={{ color: '#3370FF' }}>{text}</span>}
                    />
                    <div
                      css={{
                        fontSize: '12px',
                        color: '#8f959e',
                        overflow: 'hidden',
                        whiteSpace: 'pre-wrap',
                      }}
                    >
                      {`${item.region} - ${item.subregion}, population is ${item.population}, area is ${item.area}`}
                    </div>
                  </div>
                </div>
              ),
            }))}
          />
        </Col>
      </Row>
    </div>
  );
};

const AutoCompletion = () => {
  const [inputValue, setInputValue] = useState('');
  return (
    <Select
      value={inputValue}
      style={{ width: '200px' }}
      triggerElement={({ inputRef, onInputChange, onInternalKeyDown, setOverlayVisible }) => (
        <div>
          <Input
            ref={inputRef as RefObject<HTMLInputElement>}
            value={inputValue}
            onChange={e => {
              onInputChange(e);
              setInputValue(e.target.value);
            }}
            onKeyDown={onInternalKeyDown}
            onClick={({ target }) => {
              if (target !== inputRef.current) {
                inputRef.current?.focus();
              }
              setOverlayVisible();
            }}
          />
        </div>
      )}
      onChange={value => {
        if (value !== undefined) {
          setInputValue(value);
        }
      }}
      options={[
        {
          name: '顾乡',
          department: 'Desktop-K',
        },
        {
          name: '吴应海',
          department: 'Desktop-K',
        },
      ].map(item => ({
        value: item.name,
        labelRenderer: () => (
          <div
            css={{
              display: 'inline-flex',
              alignItems: 'center',
              width: '100%',
            }}
          >
            <Avatar css={{ marginRight: 8, flex: 'none' }} children={item.name} />
            <div
              css={{
                overflow: 'hidden',
                textOverflow: 'ellipsis',
                whiteSpace: 'nowrap',
              }}
            >
              <Highlighter
                searchWords={[inputValue]}
                textToHighlight={item.name}
                markRenderer={text => <span css={{ color: '#3370FF' }}>{text}</span>}
              />
              <div
                css={{
                  fontSize: '12px',
                  color: '#8f959e',
                  overflow: 'hidden',
                  whiteSpace: 'pre-wrap',
                }}
              >
                {item.department}
              </div>
            </div>
          </div>
        ),
      }))}
    />
  );
};

export const TriggerElement = () => (
  <>
    <AutoCompletion />
    <br />
    <br />
    <Select
      defaultValue="Jack"
      triggerElement={({ valueSet, inputRef, onInternalKeyDown, onInternalKeyUp, setOverlayVisible }) => (
        <Button
          ref={inputRef as RefObject<HTMLButtonElement>}
          onKeyDown={onInternalKeyDown}
          onKeyUp={onInternalKeyUp}
          onClick={({ target }) => {
            if (target !== inputRef.current) {
              inputRef.current?.focus();
            }
            setOverlayVisible();
          }}
        >
          {[...valueSet][0]}
        </Button>
      )}
      options={[
        {
          value: 'Jack',
        },
        {
          value: 'Lucy',
        },
        {
          value: 'Tom',
        },
      ]}
    />
    <br />
    <br />
    <Select
      defaultValue="Jack"
      triggerElement={({ valueSet, inputRef, onInternalKeyDown, onInternalKeyUp, setOverlayVisible }) => (
        <span
          style={{ outline: 'none' }}
          tabIndex={0}
          ref={inputRef as RefObject<HTMLSpanElement>}
          onKeyDown={onInternalKeyDown}
          onKeyUp={onInternalKeyUp}
          onClick={({ target }) => {
            if (target !== inputRef.current) {
              inputRef.current?.focus();
            }
            setOverlayVisible();
          }}
        >
          {[...valueSet][0]}
        </span>
      )}
      options={[
        {
          value: 'Jack',
        },
        {
          value: 'Lucy',
        },
        {
          value: 'Tom',
        },
      ]}
    />
  </>
);

export const ControlledMode = () => {
  const [state, setState] = useState<number | undefined>(0);
  return (
    <Select value={state} style={{ width: 300 }} onChange={setState}>
      <Option value={0}>Jack</Option>
      <Option value={1}>Lucy</Option>
      <Option value={2} disabled>
        Disabled
      </Option>
      <Option value={3}>Gu Xiang</Option>
    </Select>
  );
};

export { default as Placement } from '../demo/placement';

export const inputting_test = () => (
  <div>
    <table>
      <thead>
        <tr>
          <th />
          {([false, true] as const).map(showSearch =>
            [undefined, 'multiple'].map(multiple => (
              <th key={`${showSearch}-${multiple}`}>
                {`${showSearch ? '支持搜索' : '不搜索'}${multiple === 'multiple' ? ' 多选' : ' 单选'}`}
              </th>
            )),
          )}
        </tr>
      </thead>
      <tbody>
        {([undefined, false, true] as const).map(open => (
          // eslint-disable-next-line react/jsx-key
          <tr>
            <th>{open ? '常开' : open === undefined ? '一般情况' : '常关'}</th>
            {([false, true] as const).map(showSearch =>
              ([false, true] as const).map(multiple => (
                <td key={`${open}-${showSearch}-${multiple}`}>
                  <div>
                    <p />
                    <Select
                      showSearch={showSearch}
                      open={open}
                      mode={multiple ? 'multiple' : ('single' as never)}
                      defaultValue={open === false ? (multiple ? ['Jack'] : ('Jack' as never)) : undefined}
                      style={{ width: 300 }}
                      placeholder="Select a person"
                      filterOption={(input, option) => option.value.toLowerCase().indexOf(input.toLowerCase()) >= 0}
                      options={options}
                    />
                    <p />
                  </div>
                </td>
              )),
            )}
          </tr>
        ))}
      </tbody>
    </table>
  </div>
);

export const Inputting_test_autofocus = () => {
  const selectRef = React.useRef<SelectRef>(null);
  return (
    <>
      <p>多个 autoFocus</p>

      <Select
        autoFocus
        showSearch={true}
        style={{ width: 300 }}
        placeholder="Select a person"
        filterOption={(input, option) => option.value.toLowerCase().indexOf(input.toLowerCase()) >= 0}
        options={options}
      />
      <Select
        autoFocus
        showSearch={true}
        style={{ width: 300 }}
        placeholder="Select a person"
        filterOption={(input, option) => option.value.toLowerCase().indexOf(input.toLowerCase()) >= 0}
        options={options}
      />
      <Select
        autoFocus
        showSearch={true}
        style={{ width: 300 }}
        placeholder="Select a person"
        filterOption={(input, option) => option.value.toLowerCase().indexOf(input.toLowerCase()) >= 0}
        options={options}
      />
      <p>Focus Select</p>
      <Button onClick={() => selectRef.current?.focus()}>Trigger Focus</Button>
      <Select
        ref={selectRef}
        showSearch={true}
        style={{ width: 300 }}
        placeholder="Select a person"
        filterOption={(input, option) => option.value.toLowerCase().indexOf(input.toLowerCase()) >= 0}
        options={options}
      />
    </>
  );
};
