// @universe-design/react/src/select/stories/screenshot.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C21CustomDropdownMenuSearch from '../demo/21.custom-dropdown-menu-search';
import CAll from '../demo/all';
import CCheckerScreenshot from '../demo/checker.screenshot';
import CLongTextScreenshot from '../demo/long-text.screenshot';
import CNoData from '../demo/no-data';
import CPlacement from '../demo/placement';
import CSearchValueScreenshot from '../demo/searchValue.screenshot';
import CSelectAllScreenshot from '../demo/select-all.screenshot';
import CSelectorTextOverflow from '../demo/selector-text-overflow';
import CZSearch from '../demo/z.search';
if (!(C21CustomDropdownMenuSearch as Story).storyName) {
  (C21CustomDropdownMenuSearch as Story).storyName = '21 custom dropdown menu search';
}
if (!(CAll as Story).storyName) {
  (CAll as Story).storyName = 'All';
}
if (!(CCheckerScreenshot as Story).storyName) {
  (CCheckerScreenshot as Story).storyName = 'Checker screenshot';
}
if (!(CLongTextScreenshot as Story).storyName) {
  (CLongTextScreenshot as Story).storyName = 'Long text screenshot';
}
if (!(CNoData as Story).storyName) {
  (CNoData as Story).storyName = 'No data';
}
if (!(CPlacement as Story).storyName) {
  (CPlacement as Story).storyName = 'Placement';
}
if (!(CSearchValueScreenshot as Story).storyName) {
  (CSearchValueScreenshot as Story).storyName = 'Search value screenshot';
}
if (!(CSelectAllScreenshot as Story).storyName) {
  (CSelectAllScreenshot as Story).storyName = 'Select all screenshot';
}
if (!(CSelectorTextOverflow as Story).storyName) {
  (CSelectorTextOverflow as Story).storyName = 'Selector text overflow';
}
if (!(CZSearch as Story).storyName) {
  (CZSearch as Story).storyName = 'Z search';
}

export default {
  title: 'Components/Select/截图测试',
  tags: ['autodocs'],
};

export {
  C21CustomDropdownMenuSearch,
  CAll,
  CCheckerScreenshot,
  CLongTextScreenshot,
  CNoData,
  CPlacement,
  CSearchValueScreenshot,
  CSelectAllScreenshot,
  CSelectorTextOverflow,
  CZSearch,
};
