// @universe-design/react/src/select/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Basic from '../demo/0.basic';
import C1Search from '../demo/1.search';
import C11AutomaticTokenization from '../demo/11.automatic-tokenization';
import C12SearchSort from '../demo/12.search-sort';
import C13SelectUsers from '../demo/13.select-users';
import C2Multiple from '../demo/2.multiple';
import C21CustomDropdownMenu from '../demo/21.custom-dropdown-menu';
import C22HideSelected from '../demo/22.hide-selected';
import C23BorderStyle from '../demo/23.border-style';
import C24CustomTagRender from '../demo/24.custom-tag-render';
import C25BigData from '../demo/25.big-data';
import C26ResponsiveMaxTagCount from '../demo/26.responsive-max-tag-count';
import C27CustomOptions from '../demo/27.custom-options';
import C28AutoCompletion from '../demo/28.auto-completion';
import C29SelectAll from '../demo/29.selectAll';
import C3_1EmptyStatus from '../demo/3.1.empty-status';
import C3Size from '../demo/3.size';
import C30WithAvatar from '../demo/30.with-avatar';
import C31UseRef from '../demo/31.use-ref';
import C32PopupWithSearch from '../demo/32.popup-with-search';
import C4OptionGroup from '../demo/4.option-group';
import C5OptionLabelProp from '../demo/5.option-label-prop';
import C9SearchBox from '../demo/9.search-box';
if (!(C0Basic as Story).storyName) {
  (C0Basic as Story).storyName = '0. 基础用法';
  (C0Basic as Story).parameters = {
    docs: {
      description: {
        story: '简单的用法。\n\n',
      },
    },
  };
}
if (!(C1Search as Story).storyName) {
  (C1Search as Story).storyName = '1. 支持搜索';
  (C1Search as Story).parameters = {
    docs: {
      description: {
        story:
          '可以对展开的选项进行搜索。如果不提供`option.label`或`option.labelRenderer`，我们会使用内置的`Highlighter`高亮组件对选项值进行高亮。你也可以通过`option.labelRenderer`自定义高亮逻辑。\n\n',
      },
    },
  };
}
if (!(C11AutomaticTokenization as Story).storyName) {
  (C11AutomaticTokenization as Story).storyName = '11. 自动分词';
  (C11AutomaticTokenization as Story).parameters = {
    docs: {
      description: {
        story: '试下复制`a10,b11`到输入框里。只在多选模式下可用。\n\n',
      },
    },
  };
}
if (!(C12SearchSort as Story).storyName) {
  (C12SearchSort as Story).storyName = '12. 带排序的搜索';
  (C12SearchSort as Story).parameters = {
    docs: {
      description: {
        story: '在搜索模式下对过滤结果项进行排序。\n\n',
      },
    },
  };
}
if (!(C13SelectUsers as Story).storyName) {
  (C13SelectUsers as Story).storyName = '12. 搜索用户';
  (C13SelectUsers as Story).parameters = {
    docs: {
      description: {
        story: '一个带有远程搜索，防抖控制，请求时序控制，加载状态的多选示例。\n\n',
      },
    },
  };
}
if (!(C2Multiple as Story).storyName) {
  (C2Multiple as Story).storyName = '2. 多选';
  (C2Multiple as Story).parameters = {
    docs: {
      description: {
        story: '从已有条目中选择多条数据，多选默认标签会换行。\n\n',
      },
    },
  };
}
if (!(C21CustomDropdownMenu as Story).storyName) {
  (C21CustomDropdownMenu as Story).storyName = '21. 扩展菜单';
  (C21CustomDropdownMenu as Story).parameters = {
    docs: {
      description: {
        story: '使用`dropdownRender`对下拉菜单进行自由扩展。\n\n',
      },
    },
  };
}
if (!(C22HideSelected as Story).storyName) {
  (C22HideSelected as Story).storyName = '22. 隐藏已选择选项';
  (C22HideSelected as Story).parameters = {
    docs: {
      description: {
        story: '隐藏下拉列表中已选择的选项。\n\n',
      },
    },
  };
}
if (!(C23BorderStyle as Story).storyName) {
  (C23BorderStyle as Story).storyName = '23. 边框样式';
  (C23BorderStyle as Story).parameters = {
    docs: {
      description: {
        story: '另外 2 种边框样式。\n\n',
      },
    },
  };
}
if (!(C24CustomTagRender as Story).storyName) {
  (C24CustomTagRender as Story).storyName = '24. 自定义选择标签';
  (C24CustomTagRender as Story).parameters = {
    docs: {
      description: {
        story: '支持自定义选择标签的样式。\n\n',
      },
    },
  };
}
if (!(C25BigData as Story).storyName) {
  (C25BigData as Story).storyName = '25. 大数据';
  (C25BigData as Story).parameters = {
    docs: {
      description: {
        story: '选择器使用了虚拟滚动技术，因而获得了更好的性能。\n\n',
      },
    },
  };
}
if (!(C26ResponsiveMaxTagCount as Story).storyName) {
  (C26ResponsiveMaxTagCount as Story).storyName = '26. 溢出收起';
  (C26ResponsiveMaxTagCount as Story).parameters = {
    docs: {
      description: {
        story: '多选下通过响应式布局让选项自动收缩。该功能对性能有所消耗，不推荐在大表单场景下使用。\n\n',
      },
    },
  };
}
if (!(C27CustomOptions as Story).storyName) {
  (C27CustomOptions as Story).storyName = '27. 自定义选中后的 options 样式';
  (C27CustomOptions as Story).parameters = {
    docs: {
      description: {
        story: '开发者可以自己通过受控的 value 来更改 option 渲染的逻辑\n\n',
      },
    },
  };
}
if (!(C28AutoCompletion as Story).storyName) {
  (C28AutoCompletion as Story).storyName = '28. 受控状态自动补全';
  (C28AutoCompletion as Story).parameters = {
    docs: {
      description: {
        story: '用户可通过受控模式控制 searchValue 等装填来达到自动补全的效果\n\n',
      },
    },
  };
}
if (!(C29SelectAll as Story).storyName) {
  (C29SelectAll as Story).storyName = '29. 全选';
  (C29SelectAll as Story).parameters = {
    docs: {
      description: {
        story: '在多选模式下，可使用 `showSelectAll` 开启全选功能。\n',
      },
    },
  };
}
if (!(C3_1EmptyStatus as Story).storyName) {
  (C3_1EmptyStatus as Story).storyName = '3.1. 空状态说明';
  (C3_1EmptyStatus as Story).parameters = {
    docs: {
      description: {
        story:
          'Select 三兄弟都具有 **「空选项 | noOptions」** 和 **「空搜索结果 | notFound」** 两种空状态，以及一种加载态。\n\n- **「空选项 | noOptions」** 指没有传入或传入的选项列表为空。注意，如果可搜索，我们默认不会展示 **「空选项 | noOptions」** 空状态。\n- **「空搜索结果 | notFound」** 指开启搜索能力时，没有筛选结果。\n\n优先级：**「加载态 | loading」** > **「空选项 | noOptions」** > **「空搜索结果 | notFound」** > **「自定义下拉 | dropdownRender」**。\n\n',
      },
    },
  };
}
if (!(C3Size as Story).storyName) {
  (C3Size as Story).storyName = '3. 三种大小';
  (C3Size as Story).parameters = {
    docs: {
      description: {
        story: '支持 3 种大小，分别为`large`、`medium`、`small`，默认为 `medium`。\n\n',
      },
    },
  };
}
if (!(C30WithAvatar as Story).storyName) {
  (C30WithAvatar as Story).storyName = '30. 带 Avatar 的选择';
  (C30WithAvatar as Story).parameters = {
    docs: {
      description: {
        story: '`label` 支持传递 ReactNode, 所以能够实现选项中嵌入 Avatar 的功能。\n',
      },
    },
  };
}
if (!(C31UseRef as Story).storyName) {
  (C31UseRef as Story).storyName = '31. 使用 Ref 上的一些方法';
  (C31UseRef as Story).parameters = {
    docs: {
      description: {
        story:
          '可以通过 ref 直接操纵 Select。这里来借助 `scrollIntoView` 方法实现将命中搜索条件的选项滚动到可视区域内。（其他方法参考[SelectRef](#SelectRef)）\n',
      },
    },
  };
}
if (!(C32PopupWithSearch as Story).storyName) {
  (C32PopupWithSearch as Story).storyName = '32. popup 里集成搜索';
  (C32PopupWithSearch as Story).parameters = {
    docs: {
      description: {
        story: '可以通过 `dropdownRenderer` 在 popup 里面集成搜索功能。\n',
      },
    },
  };
}
if (!(C4OptionGroup as Story).storyName) {
  (C4OptionGroup as Story).storyName = '4. 选项分组';
  (C4OptionGroup as Story).parameters = {
    docs: {
      description: {
        story: '用`OptGroup`进行选项分组。\n\n',
      },
    },
  };
}
if (!(C5OptionLabelProp as Story).storyName) {
  (C5OptionLabelProp as Story).storyName = '5. 定制回填内容';
  (C5OptionLabelProp as Story).parameters = {
    docs: {
      description: {
        story:
          '使用`optionLabelProp`指定回填到选择框的`Option`属性。\n\n\n<style>\n.demo-option-label-item > span {\n  margin-right: 6px;\n}\n</style>\n',
      },
    },
  };
}
if (!(C9SearchBox as Story).storyName) {
  (C9SearchBox as Story).storyName = '9. 远程搜索';
  (C9SearchBox as Story).parameters = {
    docs: {
      description: {
        story: '搜索远程数据。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Select/官网示例',
  tags: ['autodocs'],
};

export {
  C0Basic,
  C1Search,
  C11AutomaticTokenization,
  C12SearchSort,
  C13SelectUsers,
  C2Multiple,
  C21CustomDropdownMenu,
  C22HideSelected,
  C23BorderStyle,
  C24CustomTagRender,
  C25BigData,
  C26ResponsiveMaxTagCount,
  C27CustomOptions,
  C28AutoCompletion,
  C29SelectAll,
  C3_1EmptyStatus,
  C3Size,
  C30WithAvatar,
  C31UseRef,
  C32PopupWithSearch,
  C4OptionGroup,
  C5OptionLabelProp,
  C9SearchBox,
};
