// @universe-design/react/lib/color-picker/components/ColorPickerHue.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ColorPickerHue = void 0;
var _u1u1p7mj = _interopRequireDefault(require("./../../.stylex/u1u1p7mj"));
var _u5shkri = _interopRequireDefault(require("./../../.stylex/u5shkri"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _react2 = require("@universe-design/stylex/react");
var _clsx = _interopRequireDefault(require("clsx"));
var _clamp = _interopRequireDefault(require("lodash/clamp"));
var _ColorPickerContext = require("../ColorPickerContext");
var _util = require("../util");
var _type = require("../type");
//.u1euv43f{position:absolute;}
//.ucv7f5c{top:-4px;}
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const ColorPickerHue = exports.ColorPickerHue = /*#__PURE__*/_react.default.forwardRef((_ref, ref) => {
  let {
    className,
    onHueDragEnd,
    ...props
  } = _ref;
  const setValue = (0, _ColorPickerContext.useColorPickerContextSelector)(o => o.setValue);
  const rootCssScope = (0, _ColorPickerContext.useColorPickerContextSelector)(o => o.rootCssScope);
  const transformDisplayColor = (0, _ColorPickerContext.useColorPickerContextSelector)(o => o.transformDisplayColor);
  const displayColor = (0, _ColorPickerContext.useColorPickerContextSelector)(o => o.displayColor);
  const hsv = (0, _ColorPickerContext.useColorPickerContextSelector)(o => o.hsv);
  const prefix = rootCssScope('-hue');
  const sliderHandlerRef = _react.default.useRef(null);
  const calculateChange = event => {
    const el = sliderHandlerRef.current;
    if (!el) {
      return 0;
    }
    const containerWidth = el.clientWidth;
    const distance = event.pageX - (el.getBoundingClientRect().left + window.pageXOffset);
    const val = (0, _clamp.default)(distance / containerWidth, 0, 1);
    return Math.round(val * 100);
  };
  const handleChange = event => {
    const value = calculateChange(event);
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    const h = (0, _clamp.default)(value * 360 / 100, 0, 360);
    setValue({
      ...hsv,
      h
    }, {
      source: _type.ColorPickerChangeSource.HUE
    });
  };
  const handlePointerUp = e => {
    onHueDragEnd === null || onHueDragEnd === void 0 ? void 0 : onHueDragEnd(e);
    window.removeEventListener('pointermove', handleChange);
    window.removeEventListener('pointerup', handlePointerUp);
  };

  // eslint-disable-next-line @typescript-eslint/no-magic-numbers
  const left = `${hsv.h * 100 / 360}%`;
  const sliderBackground = _react.default.useMemo(() => `linear-gradient(to right, ${transformDisplayColor((0, _util.formatColor)('#f00')).rgbString} 0%, ${transformDisplayColor((0, _util.formatColor)('#ff0')).rgbString} 17%, ${transformDisplayColor((0, _util.formatColor)('#0f0')).rgbString} 33%, ${transformDisplayColor((0, _util.formatColor)('#0ff')).rgbString} 50%, ${transformDisplayColor((0, _util.formatColor)('#00f')).rgbString} 67%, ${transformDisplayColor((0, _util.formatColor)('#f0f')).rgbString} 83%, ${transformDisplayColor((0, _util.formatColor)('#f00')).rgbString} 100%)`, [transformDisplayColor]);
  return (0, _react2.jsx)("div", (0, _extends2.default)({}, props, {
    ref: ref,
    className: (0, _clsx.default)(prefix, className)
  }), (0, _react2.jsx)("div", {
    className: (0, _clsx.default)(`${prefix}-display`, `${prefix}-display-highlight`),
    style: {
      background: displayColor && displayColor.rgbString
    }
  }), (0, _react2.jsx)("div", {
    className: `${prefix}-slider`,
    ref: sliderHandlerRef,
    onPointerDown: event => {
      handleChange(event.nativeEvent);
      window.addEventListener('pointermove', handleChange);
      window.addEventListener('pointerup', handlePointerUp);
    },
    style: {
      background: sliderBackground
    }
  }, (0, _react2.jsx)("div", {
    className: `${prefix}-pointer`,
    style: {
      left
    }
  }, (0, _react2.jsx)("svg", {
    width: 12,
    height: 5,
    sx: [{
      "props": {
        "qhf8xq": _u1u1p7mj.default,
        "Bhzewxz": _u5shkri.default
      }
    }]
  }, (0, _react2.jsx)("path", {
    d: "M0,4 V5 H10 V4 L5,0 Z",
    style: {
      fill: 'currentcolor'
    }
  })))));
});
