// @universe-design/react/lib/color-picker/components/ColorPickerColorBlock.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ColorPickerColorBlock = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _DoneOutlined = _interopRequireDefault(require("@universe-design/icons-react/DoneOutlined"));
var _styles = require("@universe-design/styles");
var _allLight = require("@universe-design/styles/cjs/all.light.atom");
var _allDark = require("@universe-design/styles/cjs/all.dark.atom");
var _type = require("../type");
var _util = require("../util");
var _hooks = require("../../config-provider/hooks");
var _ColorPickerContext = require("../ColorPickerContext");
var _util2 = require("../../util");
/**
 * @public
 * `<ColorPickerColorBlock />` 组件的属性。
 */

const ColorPickerColorBlock = exports.ColorPickerColorBlock = /*#__PURE__*/_react.default.forwardRef((_ref, ref) => {
  let {
    color,
    className,
    onClick,
    ...restProps
  } = _ref;
  /**
   * Context
   */
  const {
    value,
    setValue,
    showDarkCheckmark,
    rootCssScope,
    transformDisplayColor
  } = (0, _ColorPickerContext.useColorPickerContext)();
  const prefix = rootCssScope('-swatches');

  /**
   * Theme
   */
  const theme = (0, _hooks.useTheme)();
  const {
    'bg-float': panelBgColor = theme === 'light' ? _allLight.bg_float : _allDark.bg_float
  } = (0, _styles.useTokens)();
  const colorObj = _react.default.useMemo(() => (0, _util2.isNonNullable)(color) ? (0, _util.formatColor)(color) : color, [color]);
  const displayColor = _react.default.useMemo(() => (0, _util2.isNonNullable)(colorObj) ? transformDisplayColor(colorObj) : undefined, [colorObj, transformDisplayColor]);

  /**
   * Handler
   */
  const handleClick = _react.default.useCallback(ev => {
    (0, _util2.isFunction)(onClick) && onClick(ev);
    if ((0, _util2.isNonNullable)(colorObj === null || colorObj === void 0 ? void 0 : colorObj.rgb)) {
      setValue(colorObj === null || colorObj === void 0 ? void 0 : colorObj.rgb, {
        source: _type.ColorPickerChangeSource.COLOR_BLOCK
      });
      return;
    }
    setValue(undefined, {
      source: _type.ColorPickerChangeSource.COLOR_BLOCK
    });
  }, [colorObj === null || colorObj === void 0 ? void 0 : colorObj.rgb, onClick, setValue]);

  /**
   * Computed state
   */
  const isActive = value && (colorObj === null || colorObj === void 0 ? void 0 : colorObj.color.equals(value));
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({}, restProps, {
    ref: ref,
    className: (0, _clsx.default)(`${prefix}-cell-wrapper`, isActive && `${prefix}-cell-active`, className),
    onClick: handleClick
  }), displayColor ? /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(`${prefix}-row-cell`, displayColor.color.equals(panelBgColor) && `${prefix}-row-cell-highlight`, showDarkCheckmark(displayColor.hex, displayColor) && `${prefix}-row-cell-icon-deepen`),
    style: {
      backgroundColor: displayColor.rgbString
    }
  }, isActive ? /*#__PURE__*/_react.default.createElement(_DoneOutlined.default, null) : null) : /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(`${prefix}-row-cell`, `${prefix}-row-cell-highlight`, `${prefix}-row-cell-empty`)
  }, /*#__PURE__*/_react.default.createElement("svg", {
    width: "100%",
    height: "100%",
    viewBox: "0 0 16 16"
  }, /*#__PURE__*/_react.default.createElement("path", {
    d: "m 0 0l 16 16",
    stroke: "currentColor"
  }))));
});
