// @universe-design/react/lib/color-picker/components/ColorInput.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ColorInput = void 0;
var _react = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _clamp = _interopRequireDefault(require("lodash/clamp"));
var _hooks = require("../../config-provider/hooks");
var _type = require("../type");
var _Input = _interopRequireDefault(require("../../shared/components/Input"));
const ColorInput = props => {
  const {
    colorKey,
    color,
    onChange,
    prefixCls
  } = props;
  const [value, setValue] = (0, _react.useState)(color);
  const cssScope = (0, _hooks.useCssScope)('color-picker-input');
  const prefix = prefixCls || cssScope();
  (0, _react.useEffect)(() => {
    setValue(color);
  }, [color]);
  const colorVal = (0, _react.useMemo)(() => `${value}`.toUpperCase(), [value]);
  const judgeVal = key => {
    const val = value.toString();
    if (key !== _type.HrgbKey.HEX) {
      if (/^-?[0-9]+\.?[0-9]*$/.test(val)) {
        // eslint-disable-next-line @typescript-eslint/no-magic-numbers
        return (0, _clamp.default)(parseInt(val, 10), 0, 255);
      }
      return color;
    } else {
      const matchGroup = /^#?([0-9A-Fa-f]{0,6})$/.exec(val);
      if (!matchGroup) {
        return color.toString().toUpperCase();
      }
      const colorStr = matchGroup[1].toUpperCase();
      const len = colorStr.length;
      if (len !== 3) {
        return colorStr + '0'.repeat(6 - len);
      } else {
        let newVal = '';
        for (let i = 0; i < len; i++) {
          newVal += colorStr[i].repeat(2);
        }
        return newVal;
      }
    }
  };
  const handleBlur = key => {
    try {
      const newVal = judgeVal(key);
      setValue(newVal);
      onChange && onChange(key, newVal);
    } catch (err) {
      /* empty */
    }
  };
  const handeKeyUp = (e, key) => {
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    if (e.keyCode === 13) {
      handleBlur(key);
    }
  };
  return /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(prefix, colorKey === _type.HrgbKey.HEX && `${prefix}-hex`)
  }, /*#__PURE__*/_react.default.createElement("span", {
    className: `${prefix}-title`
  }, colorKey.toUpperCase()), /*#__PURE__*/_react.default.createElement("div", {
    className: `${prefix}-outer`
  }, /*#__PURE__*/_react.default.createElement(_Input.default, {
    type: "text",
    className: `${prefix}-inner`,
    value: colorVal,
    onChange: e => {
      setValue(e.target.value);
    },
    onBlur: () => handleBlur(colorKey),
    onKeyUp: e => handeKeyUp(e, colorKey)
  })));
};
exports.ColorInput = ColorInput;
