// @universe-design/react/lib/color-picker/components/ColorPickerSaturation.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ColorPickerSaturation = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _clamp = _interopRequireDefault(require("lodash/clamp"));
var _clsx = _interopRequireDefault(require("clsx"));
var _ColorPickerContext = require("../ColorPickerContext");
var _composeRefs = _interopRequireDefault(require("../../util/composeRefs"));
var _util = require("../../util");
var _type = require("../type");
var _util2 = require("../util");
/**
 * @public
 */

/**
 * 调整饱和度组件
 */
const ColorPickerSaturation = exports.ColorPickerSaturation = /*#__PURE__*/_react.default.forwardRef((_ref, ref) => {
  let {
    className,
    style,
    onMouseDown,
    onPointerDown,
    onSaturationDragEnd,
    ...props
  } = _ref;
  const setValue = (0, _ColorPickerContext.useColorPickerContextSelector)(o => o.setValue);
  const rootCssScope = (0, _ColorPickerContext.useColorPickerContextSelector)(o => o.rootCssScope);
  const displayColor = (0, _ColorPickerContext.useColorPickerContextSelector)(o => o.displayColor);
  const transformDisplayColor = (0, _ColorPickerContext.useColorPickerContextSelector)(o => o.transformDisplayColor);
  const hsv = (0, _ColorPickerContext.useColorPickerContextSelector)(o => o.hsv);
  const {
    hsv: displayHSV
  } = transformDisplayColor((0, _util2.formatColor)(hsv));
  const prefix = rootCssScope('-saturation');
  const divRef = (0, _react.useRef)(null);
  const mergedRef = _react.default.useMemo(() => (0, _composeRefs.default)(divRef, ref), [ref]);
  const position = {
    top: 100 - hsv.v * 100,
    left: hsv.s * 100
  };
  const getPosition = event => {
    const el = divRef.current;
    if (!el) {
      return position;
    }
    const containerRect = el.getBoundingClientRect();
    const {
      width: containerWidth,
      height: containerHeight
    } = containerRect;
    let left = event.pageX - (containerRect.left + window.pageXOffset);
    let top = event.pageY - (containerRect.top + window.pageYOffset);
    left = (0, _clamp.default)(left, 0, containerWidth);
    top = (0, _clamp.default)(top, 0, containerHeight);
    return {
      left: left * 100 / containerWidth,
      top: top * 100 / containerHeight
    };
  };
  const handleChange = event => {
    const pos = getPosition(event);
    setValue({
      h: hsv.h,
      s: pos.left / 100,
      v: 1 - pos.top / 100,
      a: hsv.a
    }, {
      source: _type.ColorPickerChangeSource.SATURATION
    });
  };
  const handlePointerUp = e => {
    onSaturationDragEnd === null || onSaturationDragEnd === void 0 ? void 0 : onSaturationDragEnd(e);
    window.removeEventListener('pointermove', handleChange);
    window.removeEventListener('pointerup', handlePointerUp);
  };
  const bindEventListeners = () => {
    window.addEventListener('pointermove', handleChange);
    window.addEventListener('pointerup', handlePointerUp);
  };
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({}, props, {
    className: (0, _clsx.default)(prefix, className),
    style: {
      ...style,
      background: displayColor && `hsl(${displayHSV.h}, 100%, 50%)`
    },
    ref: mergedRef,
    onPointerDown: event => {
      (0, _util.isFunction)(onMouseDown) && onMouseDown(event);
      (0, _util.isFunction)(onPointerDown) && onPointerDown(event);
      handleChange(event.nativeEvent);
      bindEventListeners();
    }
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: `${prefix}-border`
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: `${prefix}-white`
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: `${prefix}-black`
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(`${prefix}-circle`
    // 暂时先去掉这个 circle 的 judge，全部显示白色就行，这个和 https://www.google.com.hk/search?q=color+picker 的行为一致
    // judgeNeedDeepenIcon(hex) && `${prefix}-circle-deepen`
    ),
    style: {
      top: `${position.top}%`,
      left: `${position.left}%`,
      background: displayColor && displayColor.rgbString
    }
  })));
});
