// @universe-design/react/lib/color-picker/components/Panel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Panel = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _ColorInput = require("./ColorInput");
var _type = require("../type");
var _const = require("../const");
var _ColorPickerContext = require("../ColorPickerContext");
const Panel = exports.Panel = /*#__PURE__*/_react.default.forwardRef((_ref, ref) => {
  let {
    className,
    ...props
  } = _ref;
  const setValue = (0, _ColorPickerContext.useColorPickerContextSelector)(o => o.setValue);
  const rootCssScope = (0, _ColorPickerContext.useColorPickerContextSelector)(o => o.rootCssScope);
  const rgb = (0, _ColorPickerContext.useColorPickerContextSelector)(o => o.rgb);
  const hex = (0, _ColorPickerContext.useColorPickerContextSelector)(o => o.hex);
  const prefix = rootCssScope('-panel');
  const handleChange = (key, val) => {
    if (_const.RGBKEY.includes(key)) {
      const newRgb = Object.assign({}, rgb, {
        [key]: val
      });
      setValue(newRgb, {
        source: _type.ColorPickerChangeSource.INPUT
      });
      return;
    }
    setValue(val.toString(), {
      source: _type.ColorPickerChangeSource.INPUT
    });
  };
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({}, props, {
    className: (0, _clsx.default)(prefix, className),
    ref: ref
  }), (() => _const.COLORKEY.map(key => {
    const curKeyColor = key === _type.HrgbKey.HEX ? hex : rgb[key];
    return /*#__PURE__*/_react.default.createElement(_ColorInput.ColorInput, {
      colorKey: key,
      key: key,
      color: curKeyColor,
      onChange: handleChange
    });
  }))());
});
