// node-unix-socket/js/dgram.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.DgramSocket = void 0;
var events_1 = require("events");
var addon_1 = require("./addon");
/**
 * DgramSocket is used to create a SOCK_DGRAM unix domain socket.
 * Currently DgramSocket doesn't work with `cluster` module.
 *
 * DgramSocket is also an `EventEmitter` and will emit events including:
 *
 * ### Event: `'data'`
 * - buffer `Buffer`
 * - path `string`
 *
 * Emitted when data is received. `path` indicates remote address information.
 *
 * ### Event: `'error'`
 * - error `Error`
 *
 * Emitted when an error occurs.
 *
 * ### Event: `'close'`
 * The 'close' event is emitted after a socket is closed with close().
 */
var DgramSocket = /** @class */ (function (_super) {
    __extends(DgramSocket, _super);
    function DgramSocket() {
        var _this = _super.call(this) || this;
        _this.closed = false;
        _this.onData = function (buf, filepath) {
            process.nextTick(function () {
                _this.emit('data', buf, filepath);
            });
        };
        _this.onError = function (err) {
            _this.close();
            _this.emit('error', err);
        };
        _this.emit = _this.emit.bind(_this);
        _this.wrap = new addon_1.DgramSocketWrap(_this);
        // TODO currently we can't get this object in rust side
        _this.wrap.init(_this.wrap);
        _this.wrap.startRecv();
        _this.on('_data', _this.onData);
        _this.on('_error', _this.onError);
        return _this;
    }
    DgramSocket.prototype.checkClosed = function () {
        if (this.closed) {
            throw new Error('DgramSocket has been closed');
        }
    };
    /**
     * Listen for datagram messages on a path.
     * @param socketPath
     */
    DgramSocket.prototype.bind = function (socketPath) {
        this.checkClosed();
        this.wrap.bind(socketPath);
    };
    /**
     * Send messages to the destination path.
     * @param buf
     * @param offset
     * @param length
     * @param destPath
     * @param onWrite
     */
    DgramSocket.prototype.sendTo = function (buf, offset, length, destPath, onWrite) {
        this.checkClosed();
        this.wrap.sendTo(buf, offset, length, destPath, onWrite);
    };
    /**
     * @returns the SO_RCVBUF socket receive buffer size in bytes.
     */
    DgramSocket.prototype.getRecvBufferSize = function () {
        return this.wrap.getRecvBufferSize();
    };
    /**
     * Sets the SO_RCVBUF socket option. Sets the maximum socket receive buffer in bytes.
     * @param size
     * @returns
     */
    DgramSocket.prototype.setRecvBufferSize = function (size) {
        return this.wrap.setRecvBufferSize(size);
    };
    /**
     * @returns the SO_SNDBUF socket send buffer size in bytes.
     */
    DgramSocket.prototype.getSendBufferSize = function () {
        return this.wrap.getSendBufferSize();
    };
    /**
     * Sets the SO_SNDBUF socket option. Sets the maximum socket send buffer in bytes.
     * @param size
     * @returns
     */
    DgramSocket.prototype.setSendBufferSize = function (size) {
        return this.wrap.setSendBufferSize(size);
    };
    /**
     * Returns the bound address.
     * @returns
     */
    DgramSocket.prototype.address = function () {
        return this.wrap.address();
    };
    /**
     * Close the underlying socket and stop listening for data on it.
     * @returns
     */
    DgramSocket.prototype.close = function () {
        if (this.closed) {
            return;
        }
        this.closed = true;
        this.wrap.close();
    };
    return DgramSocket;
}(events_1.EventEmitter));
exports.DgramSocket = DgramSocket;
