// node-unix-socket/js/seqpacket.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SeqpacketSocket = exports.SeqpacketServer = void 0;
var events_1 = require("events");
var addon_1 = require("./addon");
/**
 * SeqpacketServer is used to create a SOCK_SEQPACKET server.
 * Note that sockets of SOCK_SEQPACKET don't works on MacOS and currently SeqpacketServer doesn't work with `cluster` module, i.e. you can't share a SeqpacketServer across different Node.js processes.
 *
 * SeqpacketServer is also an `EventEmitter` and will emit events including:
 *
 * ### Event: `'connection'`:
 * - socket `SeqpacketSocket`
 * - bindpath `string`
 *
 * Emitted when a new connection is made.
 *
 * ### Event: `'error'`
 * - error `Error`
 *
 * Emitted when an error occurs.
 *
 * ### Event: `'close'`
 *
 * Emitted when the server closes.
 */
var SeqpacketServer = /** @class */ (function (_super) {
    __extends(SeqpacketServer, _super);
    function SeqpacketServer() {
        var _this = _super.call(this) || this;
        _this.closed = false;
        _this.onError = function (err) {
            // TODO test this
            _this.emit('error', err);
        };
        _this.onConnection = function (fd, addr) {
            var socket = new SeqpacketSocket(fd);
            _this.emit('connection', socket, addr);
        };
        _this.emit = _this.emit.bind(_this);
        _this.wrap = new addon_1.SeqpacketSocketWrap(_this);
        // TODO currently we can't get this object in rust side
        _this.wrap.init(_this.wrap);
        _this.on('_connection', _this.onConnection);
        _this.on('_error', _this.onError);
        return _this;
    }
    SeqpacketServer.prototype.checkClosed = function () {
        if (this.closed) {
            throw new Error('SeqpacketServer has been closed');
        }
    };
    /**
     * Returns the bound address.
     * @returns
     */
    SeqpacketServer.prototype.address = function () {
        this.checkClosed();
        return this.wrap.address();
    };
    /**
     * Stops the server from accepting new connections and keeps existing connections.
     *
     * This function is synchronous.
     * @returns
     */
    SeqpacketServer.prototype.close = function () {
        if (this.closed) {
            return;
        }
        this.closed = true;
        this.wrap.close();
    };
    /**
     * Start a server listening for connections on the given path. This function is synchronous.
     * @param bindpath
     * @param backlog
     */
    SeqpacketServer.prototype.listen = function (bindpath, backlog) {
        if (backlog === void 0) { backlog = 511; }
        this.checkClosed();
        this.wrap.listen(bindpath, backlog);
    };
    /**
     * Reference the server so that it will prevent Node.js process from exiting automatically.
     */
    SeqpacketServer.prototype.ref = function () {
        this.wrap.uvRefer();
    };
    /**
     * Unreference the server so that it won't prevent Node.js process from exiting automatically.
     */
    SeqpacketServer.prototype.unref = function () {
        this.wrap.uvUnrefer();
    };
    return SeqpacketServer;
}(events_1.EventEmitter));
exports.SeqpacketServer = SeqpacketServer;
/**
 * SeqpacketSocket is an abstraction of a SOCK_SEQPACKET socket.
 *
 * SeqpacketSocket is also an `EventEmitter` and will emit events including:
 *
 * ### Event: `'connect'`
 *
 * Emitted when a socket connection is successfully established.
 *
 * ### Event: `'data'`
 *
 * - buffer `Buffer`
 * Emitted when data is received. All message boundaries in incoming datagrams are preserved.
 *
 * ### Event: `'end'`
 * Emitted when the other end of the socket signals the end of transmission, thus ending the readable side of the socket.
 *
 * ### Event: `'error'`
 * - error `Error`
 * Emitted when an error occurs. The 'close' event will be called directly following this event.
 *
 * ### Event: `'close'`
 * Emitted once the socket is fully closed.
 */
var SeqpacketSocket = /** @class */ (function (_super) {
    __extends(SeqpacketSocket, _super);
    function SeqpacketSocket(fd) {
        var _this = _super.call(this) || this;
        _this.destroyed = false;
        _this.shutdown = false;
        _this.isEnd = false;
        _this.onEnd = function () {
            _this.isEnd = true;
            _this.checkClose();
        };
        _this.onShutdown = function () {
            _this.shutdown = true;
            _this.checkClose();
            if (_this.shutdownCb) {
                _this.shutdownCb();
                _this.shutdownCb = undefined;
            }
        };
        _this.onData = function (buf) {
            _this.emit('data', buf);
        };
        _this.onError = function (err) {
            process.nextTick(function () {
                _this.emit('error', err);
                _this.destroy();
            });
        };
        _this.onConnect = function () {
            _this.wrap.startRecv();
            _this.emit('connect');
            if (_this.connectCb) {
                _this.connectCb();
                _this.connectCb = undefined;
            }
        };
        _this.emit = _this.emit.bind(_this);
        _this.wrap = new addon_1.SeqpacketSocketWrap(_this, fd);
        // TODO currently we can't get this object in rust side
        _this.wrap.init(_this.wrap);
        if (fd) {
            _this.wrap.startRecv();
        }
        _this.on('_data', _this.onData);
        _this.on('end', _this.onEnd);
        _this.on('_connect', _this.onConnect);
        _this.on('_error', _this.onError);
        _this.on('_shutdown', _this.onShutdown);
        return _this;
    }
    SeqpacketSocket.prototype.checkDestroyed = function () {
        if (this.destroyed) {
            throw new Error('SeqpacketSocket has been destroyed');
        }
    };
    SeqpacketSocket.prototype.checkClose = function () {
        if (this.isEnd && this.shutdown) {
            this.destroy();
        }
    };
    // address(): string {
    //   this.checkDestroyed();
    //   return seqAddress(this);
    // }
    /**
     * Initiate a connection on a given socket.
     *
     * This function is asynchronous. When the connection is established, the 'connect' event will be emitted.
     * However, connect() will throw error synchronously if the 'serverPath' is not a valid Seqpacket server.
     * @param serverPath
     * @param connectCb
     */
    SeqpacketSocket.prototype.connect = function (serverPath, connectCb) {
        this.checkDestroyed();
        this.connectCb = connectCb;
        this.wrap.connect(serverPath);
    };
    /**
     * Sends data on the socket. The `cb` is called when data is written to operating system.
     * @param buf
     * @param offset
     * @param length
     * @param cb
     */
    SeqpacketSocket.prototype.write = function (buf, offset, length, cb) {
        if (arguments.length === 1) {
            offset = 0;
            length = buf.length;
        }
        this.checkDestroyed();
        var v = offset || 0;
        this.wrap.write(buf, offset || 0, length || buf.length, cb);
    };
    /**
     * Half-closes the socket. i.e., it sends a FIN packet. It is possible the server will still send some data.
     * @param cb
     */
    SeqpacketSocket.prototype.end = function (cb) {
        this.shutdownCb = cb;
        this.wrap.shutdownWhenFlushed();
    };
    /**
     * Return the size of buffer that SeqpacketSocket uses to receive data. The data will be truncated if the buffer size is not large enough.
     *
     * Default size is 256KB.
     * @returns
     */
    SeqpacketSocket.prototype.getInternalReadBufferSize = function () {
        return this.wrap.getReadBufSize();
    };
    /**
     * Set the size of buffer that SeqpacketSocket uses to receive data.
     *
     * @param size
     */
    SeqpacketSocket.prototype.setInternalReadBufferSize = function (size) {
        this.wrap.setReadBufSize(size);
    };
    /**
     * Reference the socket so that it will prevent Node.js process from exiting automatically.
     */
    SeqpacketSocket.prototype.ref = function () {
        this.wrap.uvRefer();
    };
    /**
     * Unreference the socket so that it won't prevent Node.js process from exiting automatically.
     */
    SeqpacketSocket.prototype.unref = function () {
        this.wrap.uvUnrefer();
    };
    /**
     * Ensures that no more I/O activity happens on this socket. Destroys the stream and closes the connection.
     */
    SeqpacketSocket.prototype.destroy = function () {
        if (this.destroyed) {
            return;
        }
        this.destroyed = true;
        this.wrap.close();
    };
    /**
     * Alias of "destory".
     */
    SeqpacketSocket.prototype.close = function () {
        return this.destroy();
    };
    /**
     * For test only
     * @ignore
     */
    SeqpacketSocket.prototype._state = function () {
        return this.wrap.state();
    };
    return SeqpacketSocket;
}(events_1.EventEmitter));
exports.SeqpacketSocket = SeqpacketSocket;
