// @universe-design/react/src/carousel/Carousel.tsx
import * as React from 'react';
import stylex from '@universe-design/stylex';
import clsx from 'clsx';
import type { UseCarouselProps } from './CarouselContext';
import { CarouselContextProvider, useCarouselContext, useCarousel } from './CarouselContext';
import { CarouselWindow } from './CarouselWindow';
import type { StyleProps } from '../util';
import type { CarouselPrevNextSwitcherExoticProps } from './CarouselSwitcher';
import { CarouselPrevSwitcher, CarouselNextSwitcher } from './CarouselSwitcher';
import type { CarouselIndicatorExoticProps } from './CarouselIndicator';
import { CarouselIndicator } from './CarouselIndicator';
import { CarouselManagerContextProvider } from './CarouselManager';
import { carouselDefaults } from './defaults';
import { FadeCarouselTransitionDescriptor } from './transition-descriptors/fade';
import { PushCarouselTransitionDescriptor } from './transition-descriptors/push';

export type CarouselSharedProps = UseCarouselProps & StyleProps;

const ComposableCarousel = ({
  children,
  className,
  style,
  ...useCarouselProps
}: React.PropsWithChildren<CarouselSharedProps>) => {
  const { manager, contextValue } = useCarousel(useCarouselProps);
  const { rootPrefixCls } = contextValue;
  return (
    <CarouselManagerContextProvider value={manager}>
      <CarouselContextProvider value={contextValue}>
        <div
          className={clsx(
            rootPrefixCls,
            stylex(
              stylex.create({
                display: 'flex',
                flexDirection: 'column',
                height: '100%',
              }),
            ),
            className,
          )}
          {...{ style }}
        >
          {children}
        </div>
      </CarouselContextProvider>
    </CarouselManagerContextProvider>
  );
};

const CarouselMask = () => {
  const { rootPrefixCls } = useCarouselContext();
  return (
    <div
      className={clsx(
        `${rootPrefixCls}__mask`,
        stylex(
          stylex.create({
            position: 'absolute',
            height: 48,
            left: 0,
            right: 0,
            bottom: 0,
            backgroundImage: `linear-gradient(0deg, ${$token['static-black-15']} 0%, transparent 80.21%)`,
          }),
        ),
      )}
    />
  );
};

/**
 * @public
 * @docgen
 */
export interface CarouselProps extends Omit<CarouselSharedProps, 'slides' | 'transitionDescriptor'> {
  // a?: CarouselAutoPlayConfig;
  /**
   * 轮播图列表
   * @defaultValue []
   */
  children?: React.ReactElement[];
  /**
   * 指示器属性
   */
  indicatorProps?: CarouselIndicatorExoticProps;
  /**
   * 指示器位置
   * @defaultValue `inner-bottom`
   */
  indicatorPosition?: 'inner-bottom' | 'outer-bottom';
  /**
   * 翻页切换器属性
   */
  switcherProps?: CarouselPrevNextSwitcherExoticProps;
  /**
   * 转场动画
   * @defaultValue `push`
   */
  transition?: 'push' | 'fade';
}

/**
 * @public
 */
export const Carousel = ({
  children,
  indicatorProps,
  indicatorPosition = carouselDefaults.indicatorPosition,
  switcherProps,
  transition = carouselDefaults.transition,
  ...sharedProps
}: CarouselProps) => (
  <ComposableCarousel
    {...sharedProps}
    slides={children}
    transitionDescriptor={(() => {
      if (transition === 'fade') {
        return new FadeCarouselTransitionDescriptor();
      }
      return new PushCarouselTransitionDescriptor();
    })()}
  >
    <CarouselWindow>
      <CarouselPrevSwitcher {...switcherProps} />
      <CarouselNextSwitcher {...switcherProps} />
      {indicatorPosition === 'inner-bottom' && (
        <>
          <CarouselMask />
          <CarouselIndicator
            theme="light"
            {...indicatorProps}
            sx={[
              stylex.create({
                position: 'absolute',
                bottom: 0,
                left: '50%',
                transform: 'translateX(-50%)',
              }),
              indicatorProps?.sx,
            ]}
          />
        </>
      )}
    </CarouselWindow>
    {indicatorPosition === 'outer-bottom' && <CarouselIndicator theme="primary" {...indicatorProps} />}
  </ComposableCarousel>
);
