// @universe-design/react/src/carousel/CarouselSlide.tsx
import stylex from '@universe-design/stylex';
import * as React from 'react';
import { Transition } from 'react-transition-group';
import clsx from 'clsx';
import { useCarouselContext } from './CarouselContext';
import type { Slide } from './transition-descriptors/utils';
import { useCarouselManagerContext } from './CarouselManager';
import { useResizeObserver } from '@universe-design/react-utilities';
import {
  getPositionAndSize,
  useIsomorphicLayoutEffect as useLayoutEffect,
  // useLogger
} from '../hooks';
import composeRefs from '../util/composeRefs';
import { forwardHTMLRef } from '../util/forwardRef';
import { createContextHook } from '@universe-design/react-utilities';
import pick from 'lodash/pick';
import omit from 'lodash/omit';

const transitionGroupClonePropKeys = ['onExited', 'in', 'appear', 'enter', 'exit'];

export const { Provider: CarouselSlideContextProvider, useContext: useCarouselSlideContext } = createContextHook<Slide>(
  'CarouselSlideContext',
  { strict: true },
);

interface CarouselSlideExoticProps {
  /**
   * 幻灯片是否占满整个视口高度
   * @defaultValue true
   */
  fullHeight?: boolean;
  /**
   * 幻灯片是否占满整个视口宽度
   * @defaultValue true
   */
  fullWidth?: boolean;
}

/**
 * @public
 */
export const CarouselSlide = forwardHTMLRef<CarouselSlideExoticProps, 'div'>((props, ref) => {
  const { children, className, fullHeight = true, fullWidth = true, ...restProps } = props;
  // https://github.com/reactjs/react-transition-group/blob/0d8a515aa59b8ba77fcabaac479e798532d4adfa/src/utils/ChildMapping.js#L88
  const transitionGroupProps = pick(restProps, transitionGroupClonePropKeys);
  const domAttrs = omit(restProps, transitionGroupClonePropKeys);
  const { groupIndex, slideIndex } = useCarouselSlideContext();
  const { duration, rootPrefixCls } = useCarouselContext();
  const manager = useCarouselManagerContext();
  const slideElRef = React.useRef<HTMLDivElement>(null);

  const slideElResizeObserverRef = useResizeObserver<HTMLDivElement>(entry => {
    const rect = getPositionAndSize(entry.target as HTMLDivElement, entry);
    // logger.info(
    //   `Carousel useResizeObserver ${JSON.stringify({
    //     groupIndex,
    //     slideIndex,
    //   })}`,
    //   {
    //     width: rect.width,
    //     height: rect.height,
    //   }
    // );
    manager.updateSlideSize({ groupIndex, slideIndex }, rect);
  });

  const composedSlideElRef = React.useMemo(
    () =>
      composeRefs(
        ref,
        slideElRef,
        slideElResizeObserverRef,
        manager.register<HTMLDivElement>({ groupIndex, slideIndex }),
      ),
    [groupIndex, manager, ref, slideElResizeObserverRef, slideIndex],
  );

  useLayoutEffect(() => {
    // 首次渲染立即存下 slide size
    const positionedEl = slideElRef.current;
    if (!positionedEl) {
      return;
    }
    // force reflow 但是没关系，因为都是 absolute 的元素
    manager.updateSlideSize({ groupIndex, slideIndex }, getPositionAndSize(positionedEl));
  }, [groupIndex, manager, slideIndex]);

  // useLogger(`CarouselSlide ${JSON.stringify({ groupIndex, slideIndex })}`);

  return (
    <Transition
      nodeRef={slideElRef}
      timeout={{
        exit: duration,
      }}
      {...transitionGroupProps}
    >
      <div
        ref={composedSlideElRef}
        data-group-index={groupIndex}
        data-slide-index={slideIndex}
        className={clsx(
          `${rootPrefixCls}__slide`,
          stylex(
            stylex.create({
              position: 'absolute',
              left: 0,
              top: 0,
              display: 'inline-block',
              boxSizing: 'border-box',
              backgroundColor: $token['text-disabled'],
            }),
            fullWidth &&
              stylex.create({
                width: '100%',
              }),
            fullHeight &&
              stylex.create({
                height: '100%',
              }),
          ),
          className,
        )}
        {...domAttrs}
      >
        {children}
      </div>
    </Transition>
  );
});

/**
 * @public
 * @docgen
 */
export type CarouselSlideProps = React.ComponentProps<typeof CarouselSlide>;
