// @universe-design/react/src/carousel/CarouselContext.tsx
import * as React from 'react';
import { createContextHook, useLatestState, useConstant } from '@universe-design/react-utilities';
import clamp from 'lodash/clamp';
import { useSyncExternalStore } from 'use-sync-external-store/shim';
import { carouselDefaults } from './defaults';
import { useCssScope } from '../config-provider';
import { isNonNullable } from '../util';
import { CarouselManager } from './CarouselManager';
import type { CarouselTransitionDescriptor, Slide } from './transition-descriptors/utils';

/**
 * ====================
 *   AutoPlay
 * ====================
 */

/**
 * @public
 */
export interface CarouselAutoPlayConfig {
  /**
   * 幻灯片切换间隔, 单位 ms
   * @defaultValue 5000
   */
  interval?: number;
  /**
   * 是否在悬停窗口时暂停自动播放
   * @defaultValue true
   */
  pauseOnHover?: boolean;
}

const getNow = () => new Date().getTime();

/**
 * @public
 */
export interface UseCarouselAutoPlayProps {
  /**
   * 是否自动播放。
   * @defaultValue false
   */
  autoPlay?: boolean | CarouselAutoPlayConfig;
}

const useCarouselAutoPlay = (
  { autoPlay = carouselDefaults.autoPlay }: UseCarouselAutoPlayProps,
  { manager, slide }: { manager: CarouselManager; slide: Slide },
) => {
  const { interval: autoPlayInterval, pauseOnHover: autoPlayPauseOnHover } = (() => {
    if (!autoPlay || autoPlay === true) {
      return carouselDefaults.autoPlayConfig;
    }
    return {
      interval: autoPlay.interval ?? carouselDefaults.autoPlayConfig.interval,
      pauseOnHover: autoPlay.pauseOnHover ?? carouselDefaults.autoPlayConfig.pauseOnHover,
    };
  })();
  const autoPlayEnabled = Boolean(autoPlay);

  const [isAutoPlayRunning, setIsAutoPlayRunning, isAutoPlayRunningRef] = useLatestState(false);
  // 切换幻灯片的时间
  const slideChangedAt = React.useRef<number>(getNow());
  const playedAt = React.useRef<number>(getNow());
  const pausedAt = React.useRef<number>(getNow());

  // play ----> pause ----[idle]----> play ----> slide changed -[animation]->-------> slide changed ...
  React.useEffect(() => {
    // slide 变化的时候需要更新所有的时间
    slide.groupIndex;
    slide.slideIndex;
    const now = getNow();
    slideChangedAt.current = now;
    playedAt.current = now;
    pausedAt.current = now;
  }, [slide.groupIndex, slide.slideIndex]);

  React.useEffect(() => {
    // 开启 autoPlay 并且非暂停状态
    if (!autoPlayEnabled || !isAutoPlayRunning) {
      return;
    }
    let t: number;
    const loop = () => {
      t = window.setTimeout(() => {
        const elapsed = pausedAt.current - slideChangedAt.current + (getNow() - playedAt.current);
        if (elapsed > autoPlayInterval) {
          manager.setSlideIndexByStep(1);
        }
        loop();
      }, 100);
    };
    loop();
    return () => {
      window.clearTimeout(t);
    };
  }, [autoPlayEnabled, autoPlayInterval, isAutoPlayRunning, manager]);

  const play = React.useCallback(() => {
    if (isAutoPlayRunningRef.current) {
      return;
    }
    playedAt.current = getNow();
    setIsAutoPlayRunning(true);
  }, [isAutoPlayRunningRef, setIsAutoPlayRunning]);

  const pause = React.useCallback(() => {
    if (!isAutoPlayRunningRef.current) {
      return;
    }
    pausedAt.current = getNow();
    setIsAutoPlayRunning(false);
  }, [isAutoPlayRunningRef, setIsAutoPlayRunning]);

  React.useEffect(() => {
    if (autoPlay) {
      play();
    }
  }, [autoPlay, play]);

  return {
    play,
    pause,
    autoPlayPauseOnHover,
  };
};

/**
 * @public
 */
export interface UseCarouselProps extends UseCarouselAutoPlayProps {
  /**
   * 轮播图列表
   * @defaultValue []
   */
  slides?: React.ReactElement[];
  /**
   * 当前图片数组索引值的默认值
   * @defaultValue 0
   */
  defaultSlideIndex?: number;
  /**
   * 当前图片数组索引值的受控值
   */
  slideIndex?: number;
  /**
   * 是否支持无限循环跳转
   * @defaultValue false
   */
  // infinite?: boolean;
  /**
   * 动画渐变时长，单位是 ms，`push` 转场默认值为 `300`，`fade` 转场默认值为 `1000`
   * @defaultValue 300
   */
  duration?: number;
  /**
   * 图片切换时的回调
   */
  onSlideIndexChange?: (slideIndex: number) => unknown;
  /**
   * 轮播图转场动画描述符
   */
  transitionDescriptor: CarouselTransitionDescriptor;
}

/**
 * Carousel hook
 */
export const useCarousel = (props: UseCarouselProps) => {
  const {
    slides = carouselDefaults.slides,
    defaultSlideIndex = carouselDefaults.defaultSlideIndex,
    slideIndex: slideIndexProp,
    // infinite = carouselDefaults.infinite,
    onSlideIndexChange,
    autoPlay,
    transitionDescriptor,
    duration = transitionDescriptor.defaults.duration,
  } = props;
  const slideCount = slides.length;
  const rootPrefixCls = useCssScope('carousel')('');
  const trackDescriptor = useConstant(() => transitionDescriptor);
  const manager = useConstant(
    () =>
      new CarouselManager({
        trackDescriptor,
        defaultSlideIndex: clamp(slideIndexProp ?? defaultSlideIndex, 0, slideCount - 1),
      }),
  );
  manager.slideCount = slideCount;
  manager.duration = duration;
  manager.onSlideIndexChange = onSlideIndexChange;
  /**
   * States
   */
  const slide = useSyncExternalStore(
    manager.subscribe,
    React.useCallback(() => manager.state.slide, [manager]),
  );

  React.useMemo(() => {
    // TODO 这里如果 onChange 的和 controlled value 不一致会 reset 回 controlled value，但是 controlled slide 引用会变，之后的逻辑不能依赖于 slide 的引用，而是需要做 deep equal
    if (isNonNullable(slideIndexProp) && slideIndexProp !== slide.slideIndex) {
      manager.setSlideIndex(slideIndexProp, { triggerChange: false });
    }
  }, [slide.slideIndex, slideIndexProp, manager]);

  /**
   * ====================
   *   AutoPlay
   * ====================
   */
  const { play, pause, autoPlayPauseOnHover } = useCarouselAutoPlay({ autoPlay }, { manager, slide });

  return {
    manager,
    contextValue: React.useMemo(
      () => ({
        rootPrefixCls,
        slide,
        slides,
        slideCount,
        duration,
        autoPlayPauseOnHover,
        play,
        pause,
      }),
      [rootPrefixCls, slide, slides, slideCount, duration, autoPlayPauseOnHover, play, pause],
    ),
  };
};

export type CarouselContextType = ReturnType<typeof useCarousel>['contextValue'];

/**
 * Carousel 组件全局状态
 */
export const { Provider: CarouselContextProvider, useContext: useCarouselContext } =
  createContextHook<CarouselContextType>('CarouselContext', {
    strict: true,
  });
