// @universe-design/react/src/carousel/CarouselManager.ts
import produce from 'immer';
import clamp from 'lodash/clamp';
import isEqual from 'lodash/isEqual';
import minBy from 'lodash/minBy';
import { BehaviorSubject } from 'rxjs';
import { createContextHook } from '@universe-design/react-utilities';
import { forceReflow } from '../util/forceReflow';
import type { Position, Size, Slide, CarouselTransitionDescriptor } from './transition-descriptors/utils';
import { getSlideFromAbsoluteIndex, getAbsoluteIndexFromSlide } from './transition-descriptors/utils';

export class CarouselManager {
  // 类 prop 属性
  slideCount = 0;
  duration = 0;
  onSlideIndexChange: ((slide: number) => unknown) | undefined = undefined;
  // subject
  private _subject: BehaviorSubject<{
    slide: Slide;
    prevSlide: Slide | null;
    windowSize: Size;
    slideSizes: (Size & { mounted: boolean })[];
    slidesToBeRendered: Slide[];
  }>;
  private _trackDescriptor: CarouselTransitionDescriptor;
  // 类 state 属性
  private _prevSlide: Slide | null = null;
  private _slide: Slide;
  private _windowSize: Size;
  private _slideSizes: (Size & { mounted: boolean })[];
  private _slidesToBeRendered: Slide[];
  // 最新值 cache
  private _slideMetaCache: Map<string, { element: HTMLElement | null }>;
  // eslint-disable-next-line func-call-spacing
  private _transitionDisposers = new Map<string, () => void>();
  private _userEventLockSet = new Set();

  constructor(params: { trackDescriptor: CarouselTransitionDescriptor; defaultSlideIndex: number }) {
    this._trackDescriptor = params.trackDescriptor;
    this._slide = {
      groupIndex: 0,
      slideIndex: params.defaultSlideIndex,
    };
    this._windowSize = { width: 0, height: 0 };
    this._slideSizes = [];
    this._slideMetaCache = new Map();
    this._slidesToBeRendered = [];
    this._subject = new BehaviorSubject(this.state);
  }

  /**
   * ====================
   *   Public Methods
   * ====================
   */
  get state() {
    return {
      slide: this._slide,
      prevSlide: this._prevSlide,
      windowSize: this._windowSize,
      slideSizes: this._slideSizes,
      slidesToBeRendered: this._slidesToBeRendered,
    };
  }

  get canUserInteract() {
    return this._userEventLockSet.size === 0;
  }

  subscribe = (onStoreChange: () => void) => {
    const subscription = this._subject.subscribe({ next: onStoreChange });
    return () => {
      subscription.unsubscribe();
    };
  };

  /**
   * 注册 slide 节点
   */
  register =
    <T extends HTMLElement>(slide: Slide) =>
    (element: T | null) => {
      if (this._getSlideMeta(slide)) {
        this._updateSlideMeta(slide, {
          element,
        });
        return;
      }
      this._setSlideMeta(slide, { element });
    };

  /**
   * 更新视口尺寸
   */
  updateWindowSize(size: Size) {
    // logger.info('manager.updateWindowSize', size);
    this._windowSize = produce(this._windowSize, draft => {
      const { width, height } = size;
      Object.assign(draft, { width, height });
    });
    this._emit();
  }

  /**
   * 更新 slide 尺寸
   */
  updateSlideSize(slide: Slide, size: Size) {
    this._slideSizes = produce(this._slideSizes, draft => {
      const { width, height } = size;
      if (!draft[slide.slideIndex]) {
        draft[slide.slideIndex] = { width, height, mounted: true };
        return;
      }
      Object.assign(draft[slide.slideIndex], { width, height, mounted: true });
    });
    // logger.info('manager.updateSlideSize', slide, this._slideSizes);
    this._emit();
  }

  /**
   * 按 slide index 来设置当前 slide
   */
  setSlideIndex(slide: number, { triggerChange = true }: { triggerChange?: boolean } = {}) {
    // logger.info('setSlideIndex', this._userEventLockSet);
    const _slide = clamp(slide, 0, this.slideCount - 1);
    const curAbsoluteIndex = getAbsoluteIndexFromSlide(this._slide, this.slideCount);
    const nextAbsoluteIndexes = [this._slide.groupIndex - 1, this._slide.groupIndex, this._slide.groupIndex + 1].map(
      groupIndex => getAbsoluteIndexFromSlide({ groupIndex, slideIndex: _slide }, this.slideCount),
    );
    const nextSlide = getSlideFromAbsoluteIndex(
      minBy(nextAbsoluteIndexes, i => Math.abs(i - curAbsoluteIndex)) as number,
      this.slideCount,
    );
    this._setSlide(nextSlide, { triggerChange });
  }

  /**
   * 根据步长来设置新的 slide
   */
  setSlideIndexByStep(step: number) {
    // logger.info('setSlideIndexByStep', this._userEventLockSet);
    const curAbsoluteIndex = getAbsoluteIndexFromSlide(this._slide, this.slideCount);
    const nextAbsoluteIndex = curAbsoluteIndex + step;
    const nextSlide = getSlideFromAbsoluteIndex(nextAbsoluteIndex, this.slideCount);
    this._setSlide(nextSlide, { triggerChange: true });
  }

  /**
   * 更新渲染的 slides 列表
   */
  updateSlidesToBeRendered() {
    const nextSlides = this._trackDescriptor.getSlidesTOBeRendered(this);
    if (isEqual(this._slidesToBeRendered, nextSlides)) {
      return;
    }
    this._slidesToBeRendered = produce(this._slidesToBeRendered, draft => {
      draft.splice(nextSlides.length - 1, this._slidesToBeRendered.length - nextSlides.length);
      for (let i = 0; i < nextSlides.length; i++) {
        if (!draft[i]) {
          draft[i] = nextSlides[i];
          continue;
        }
        Object.assign(draft[i], nextSlides[i]);
      }
    });
    this._emit();
  }

  getSlideSize(slideIndex: number) {
    return this._slideSizes[slideIndex] ?? { width: 0, height: 0 };
  }

  // 1. 直接跳转到当前状态
  // 2. 渐变跳转到当前状态
  resetNodes(options: { slick: boolean }) {
    for (const [key, { element }] of this._slideMetaCache) {
      const slide = this._unmarshalSlide(key);
      if (!element) {
        continue;
      }
      if (options.slick) {
        this._resetNodeSlickly(slide, element);
      } else {
        this._resetNode(slide, element);
      }
    }
  }

  /**
   * ====================
   *   Slide Meta Cache
   * ====================
   */

  // 获取 slide 的 cache key
  private _marshalSlide(slide: Slide) {
    return JSON.stringify(slide);
  }
  private _unmarshalSlide(key: string) {
    return JSON.parse(key) as Slide;
  }
  // 获取 slide meta 信息
  private _getSlideMeta(slide: Slide) {
    return this._slideMetaCache.get(this._marshalSlide(slide));
  }
  // 覆盖 slide meta 信息
  private _setSlideMeta(slide: Slide, params: { element: HTMLElement | null }) {
    return this._slideMetaCache.set(this._marshalSlide(slide), params);
  }
  // 更新 slide meta 信息
  private _updateSlideMeta(slide: Slide, params: { element?: HTMLElement | null; position?: Position }) {
    const slideMeta = this._getSlideMeta(slide);
    if (!slideMeta) {
      return;
    }
    this._setSlideMeta(slide, {
      ...slideMeta,
      ...params,
    });
  }

  /**
   * ====================
   *   State
   * ====================
   */
  private _setSlide(nextSlide: Slide, options: { triggerChange: boolean }) {
    this._prevSlide = this._slide;
    this._slide = produce(this._slide, draft => {
      Object.assign(draft, nextSlide);
    });
    if (this._prevSlide !== this._slide && options.triggerChange) {
      this.onSlideIndexChange?.(this._slide.slideIndex);
    }
    this._emit();
  }

  private _emit() {
    const nextState = produce(this._subject.value, draft => {
      Object.assign(draft, this.state);
    });
    if (this._subject.value !== nextState) {
      this._subject.next(this.state);
    }
  }

  private _resetNode(slide: Slide, element: HTMLElement) {
    this._trackDescriptor.onGotoDirectly({ slide, element }, this);
  }

  private _resetNodeSlickly(slide: Slide, element: HTMLElement) {
    const cacheKey = this._marshalSlide(slide);
    // lock user event
    this._userEventLockSet.add(cacheKey);

    // 这里一定要有 disposer 是因为，如果旧的 transition 还未完成，新的 transition 就开始，这时旧的 transition 结束可能会影响新 transition 的工作
    const prevDisposer = this._transitionDisposers.get(cacheKey);
    if (prevDisposer) {
      prevDisposer();
    }

    // transition
    // enter
    this._trackDescriptor.onGotoSlicklyStart({ slide, element }, this);
    // entering
    forceReflow(element);
    const enteringDisposer = this._trackDescriptor.onGotoSlicklyActive({ slide, element }, this);
    // entered
    const t = window.setTimeout(() => {
      this._trackDescriptor.onGotoSlicklyDone({ slide, element }, this);
      this._userEventLockSet.delete(cacheKey);
      // eslint-disable-next-line @typescript-eslint/no-use-before-define
      disposer(); // 这里 dispose 一下避免自定义渐变时可能存在的内存泄露
    }, this.duration);

    // set disposer
    const disposer = () => {
      enteringDisposer?.();
      window.clearTimeout(t);
    };
    this._transitionDisposers.set(cacheKey, disposer);
  }
}

export const { Provider: CarouselManagerContextProvider, useContext: useCarouselManagerContext } =
  createContextHook<CarouselManager>('CarouselManagerContext', {
    strict: true,
  });
