// @universe-design/react/src/carousel/CarouselSwitcher.tsx
import LeftOutlined from '@universe-design/icons-react/LeftOutlined';
import RightOutlined from '@universe-design/icons-react/RightOutlined';
import stylex from '@universe-design/stylex';
import clsx from 'clsx';
import * as React from 'react';
import { sxBorderRadius, sxTransitionTimingFunction } from '../stylex';
import type { DefaultSizeType, StylexProps } from '../util';
import { forwardHTMLRef } from '../util/forwardRef';
import { useCarouselContext } from './CarouselContext';
import { useCarouselManagerContext } from './CarouselManager';

interface CarouselSwitcherExoticProps {
  icon?: React.ReactElement;
  /**
   * 翻页按钮尺寸
   * @defaultValue `medium`
   */
  size?: Exclude<DefaultSizeType, 'large'>;
}

/**
 * TODO: 等 button 迁移 stylex 之后用 ud button 改写
 */
const CarouselSwitcher = forwardHTMLRef<CarouselSwitcherExoticProps & StylexProps, 'button'>(
  (
    {
      // omit
      children,
      // props
      className,
      sx,
      icon,
      size = 'medium',
      ...buttonProps
    },
    ref,
  ) => (
    <button
      ref={ref}
      type="button"
      className={clsx(
        className,
        stylex(
          stylex.create({
            display: 'flex',
            alignItems: 'center',
            justifyContent: 'center',

            color: $token['static-white'],
            backgroundColor: $token['static-black-50'],
            border: 'none',
            cursor: 'pointer',
            transitionProperty: 'background-color, color',
            transitionDuration: '0.1s, 0.1s',
            ':hover': {
              backgroundColor: $token['static-black-80'],
            },
            ':active': {
              backgroundColor: $token['static-black-80'],
            },
            ':focus-visible': {
              outline: 'none',
              backgroundColor: $token['static-black-80'],
            },
          }),
          sxBorderRadius.xl,
          sxTransitionTimingFunction.standard,
          {
            small: stylex.create({
              width: 28,
              height: 28,
              fontSize: 14,
            }),
            medium: stylex.create({
              width: 40,
              height: 40,
              fontSize: 16,
            }),
          }[size],
          sx,
        ),
      )}
      {...buttonProps}
    >
      {icon}
    </button>
  ),
);

/**
 * @public
 * @docgen
 */
export type CarouselPrevNextSwitcherExoticProps = Omit<CarouselSwitcherExoticProps, 'icon'>;

export const CarouselPrevSwitcher = forwardHTMLRef<CarouselPrevNextSwitcherExoticProps & StylexProps, 'button'>(
  (
    {
      // omit
      children,
      // props
      onClick,
      sx,
      ...buttonProps
    },
    ref,
  ) => {
    const { rootPrefixCls } = useCarouselContext();
    const manager = useCarouselManagerContext();
    return (
      <CarouselSwitcher
        ref={ref}
        icon={<LeftOutlined />}
        className={`${rootPrefixCls}__prev-switcher`}
        {...buttonProps}
        sx={[
          stylex.create({
            position: 'absolute',
            left: 24,
            top: '50%',
            transform: 'translateY(-50%)',
          }),
          sx,
        ]}
        onClick={ev => {
          onClick?.(ev);
          if (!manager.canUserInteract) {
            return;
          }
          manager.setSlideIndexByStep(-1);
        }}
      />
    );
  },
);

export const CarouselNextSwitcher = forwardHTMLRef<CarouselPrevNextSwitcherExoticProps & StylexProps, 'button'>(
  (
    {
      // omit
      children,
      // props
      onClick,
      sx,
      ...buttonProps
    },
    ref,
  ) => {
    const { rootPrefixCls } = useCarouselContext();
    const manager = useCarouselManagerContext();
    return (
      <CarouselSwitcher
        ref={ref}
        icon={<RightOutlined />}
        className={`${rootPrefixCls}__next-switcher`}
        {...buttonProps}
        sx={[
          stylex.create({
            position: 'absolute',
            right: 24,
            top: '50%',
            transform: 'translateY(-50%)',
          }),
        ]}
        onClick={ev => {
          onClick?.(ev);
          if (!manager.canUserInteract) {
            return;
          }
          manager.setSlideIndexByStep(1);
        }}
      />
    );
  },
);
