// @universe-design/react/src/carousel/CarouselIndicator.tsx
import * as React from 'react';
import stylex from '@universe-design/stylex';
import { forwardHTMLRef } from '../util/forwardRef';
import { useCarouselContext } from './CarouselContext';
import { sxBorderRadius, sxTransitionTimingFunction } from '../stylex';
import { Space } from '../space';
import type { StylexProps } from '../util';
import clsx from 'clsx';
import { useCarouselManagerContext } from './CarouselManager';

/**
 * @public
 * @docgen
 */
export interface CarouselIndicatorExoticProps extends StylexProps {
  /**
   * 指示器类型
   * @defaultValue `dot`
   */
  type?: 'dot' | 'line';
  /**
   * 指示器主题
   * @defaultValue `light`
   */
  theme?: 'light' | 'primary';
}

export const CarouselIndicator = forwardHTMLRef<CarouselIndicatorExoticProps, 'div'>((props, ref) => {
  const { className, type = 'dot', theme = 'light', sx, ...restProps } = props;
  const { slide, slideCount, duration, rootPrefixCls } = useCarouselContext();
  const manager = useCarouselManagerContext();
  const spacing = 6;
  return (
    <Space
      className={clsx(
        `${rootPrefixCls}__indicator`,
        stylex(
          stylex.create({
            justifyContent: 'center',
            padding: '20px 0',
            flexShrink: 0,
          }),
          sx,
        ),
        className,
      )}
      size={spacing}
      ref={ref}
      inline={false}
      {...restProps}
    >
      {Array(slideCount)
        .fill(0)
        .map((_, i) => (
          <div
            key={i}
            style={{
              transitionDuration: `${duration}ms`,
            }}
            className={clsx(
              `${rootPrefixCls}__indicator-item`,
              `${rootPrefixCls}__indicator-item-${theme}`,
              `${rootPrefixCls}__indicator-item-${type}`,
              slide.slideIndex === i && `${rootPrefixCls}__indicator-item-active`,
              stylex(
                stylex.create({
                  position: 'relative',
                  transitionProperty: 'all',
                  cursor: 'pointer',
                }),
                sxBorderRadius.xl,
                sxTransitionTimingFunction.standard,
                slide.slideIndex === i
                  ? [
                      theme === 'primary'
                        ? stylex.create({
                            backgroundColor: $token['primary-fill-default'],
                          })
                        : stylex.create({
                            backgroundColor: $token['primary-on-primary-fill'],
                          }),
                      type === 'dot'
                        ? stylex.create({
                            width: 12,
                            height: 6,
                          })
                        : stylex.create({
                            width: 24,
                            height: 4,
                          }),
                    ]
                  : [
                      theme === 'primary'
                        ? stylex.create({
                            backgroundColor: $token['line-border-component'],
                            ':hover': {
                              backgroundColor: $token['primary-pri-200'],
                            },
                          })
                        : stylex.create({
                            backgroundColor: $token['static-white-30'],
                            ':hover': {
                              backgroundColor: $token['static-white-60'],
                            },
                          }),
                      type === 'dot'
                        ? stylex.create({
                            width: 6,
                            height: 6,
                          })
                        : stylex.create({ width: 24, height: 4 }),
                    ],
              ),
            )}
            onClick={() => {
              if (!manager.canUserInteract) {
                return;
              }
              manager.setSlideIndex(i);
            }}
          >
            {/* 扩大热区 */}
            <div
              className={stylex(
                stylex.create({
                  position: 'absolute',
                  left: -spacing / 2,
                  right: -spacing / 2,
                  top: -spacing / 2,
                  bottom: -spacing / 2,
                }),
              )}
            />
          </div>
        ))}
    </Space>
  );
});
