// @universe-design/react/es/progress/Circle.js
import React from 'react';
import SucceedFilled from '@universe-design/icons-react/SucceedFilled';
import ErrorFilled from '@universe-design/icons-react/ErrorFilled';
import { useCssScope } from '../config-provider/hooks';
import clsx from 'clsx';
import { ProgressStatusEnum } from './shared';
import { isFunction } from '../util';
function getPathStyles(percent, width, strokeWidth) {
  const radius = (width - strokeWidth) / 2;
  const len = Math.PI * 2 * radius;
  const gapDegree = 0;
  return {
    pathString: `
  M ${width / 2} ${width / 2}
  m 0 ${-radius}

  A ${radius} ${radius} 0 0 1 ${width / 2} ${width - strokeWidth / 2}
  A ${radius} ${radius} 0 0 1 ${width / 2} ${strokeWidth / 2}
  `,
    pathStyle: {
      strokeDasharray: `${percent / 100 * (len - gapDegree)}px ${len}px`,
      strokeDashoffset: `-${gapDegree / 2}px`
    }
  };
}
const Circle = props => {
  const {
    percent,
    showInfo = true,
    infoPosition = 'left',
    strokeWidth = 2,
    width = 16,
    status,
    strokeColor,
    trailColor,
    format
  } = props;
  const cssScope = useCssScope('progress');
  const [success, exception, normal] = ProgressStatusEnum;
  const circleStatus = status || normal;
  const {
    pathString,
    pathStyle
  } = getPathStyles(percent, width, strokeWidth);
  const renderPercent = () => {
    if (isFunction(format)) {
      return format(percent);
    }
    return `${percent}%`;
  };
  const renderSteps = () => {
    const cls = clsx([cssScope('--circle--steps'), {
      [cssScope('--circle--steps--left')]: infoPosition === 'left',
      [cssScope('--circle--steps--right')]: infoPosition === 'right'
    }]);
    return /*#__PURE__*/React.createElement("span", {
      className: cls
    }, renderPercent());
  };
  return /*#__PURE__*/React.createElement(React.Fragment, null, showInfo && infoPosition === 'left' && renderSteps(), (() => /*#__PURE__*/React.createElement("span", {
    style: {
      position: 'relative',
      display: 'flex'
    }
  }, /*#__PURE__*/React.createElement("svg", {
    width: width,
    height: width,
    fill: "none",
    color: strokeColor,
    viewBox: `0 0 ${width} ${width}`
  }, /*#__PURE__*/React.createElement("path", {
    d: pathString,
    fill: "none",
    stroke: trailColor,
    strokeWidth: strokeWidth,
    strokeLinecap: "round",
    className: clsx([cssScope('--circle--path--bg'), {
      [`${cssScope('--circle--path--success')}`]: circleStatus === success,
      [`${cssScope('--circle--path--exception')}`]: circleStatus === exception
    }])
  }), /*#__PURE__*/React.createElement("path", {
    d: pathString,
    fill: "none",
    stroke: strokeColor,
    strokeLinecap: percent === 0 ? 'butt' : 'round',
    strokeWidth: strokeWidth,
    style: pathStyle,
    className: clsx([cssScope('--circle--path'), {
      [`${cssScope('--circle--path--success')}`]: circleStatus === success,
      [`${cssScope('--circle--path--exception')}`]: circleStatus === exception
    }])
  })), /*#__PURE__*/React.createElement("span", {
    className: clsx([cssScope('--circle--icon'), cssScope('--circle--icon--success')]),
    style: {
      opacity: circleStatus === success ? 1 : 0,
      fontSize: `${width}px`
    }
  }, /*#__PURE__*/React.createElement(SucceedFilled, {
    style: {
      verticalAlign: '10px'
    }
  })), /*#__PURE__*/React.createElement("span", {
    className: clsx([cssScope('--circle--icon'), cssScope('--circle--icon--exception')]),
    style: {
      opacity: circleStatus === exception ? 1 : 0,
      fontSize: `${width}px`
    }
  }, /*#__PURE__*/React.createElement(ErrorFilled, {
    style: {
      verticalAlign: 0
    }
  }))))(), showInfo && infoPosition === 'right' && renderSteps());
};
export default Circle;
