// @universe-design/react/es/progress/Line.js
import React from 'react';
import { useCssScope } from '../config-provider/hooks';
import clsx from 'clsx';
import Steps from './Steps';
const Line = props => {
  const {
    percent,
    strokeColor = '',
    strokeWidth = 4,
    status,
    trailColor,
    showInfo = true,
    infoPosition = 'right',
    format
  } = props;
  let lineStatus = status || 'normal';
  if (lineStatus === 'normal' && percent >= 100) {
    lineStatus = 'success';
  }
  const cssScope = useCssScope('progress');
  const renderSteps = () => /*#__PURE__*/React.createElement("span", {
    className: cssScope('--steps')
  }, /*#__PURE__*/React.createElement(Steps, {
    percent: percent,
    status: lineStatus,
    format: format
  }));
  const clsBg = clsx([cssScope(), cssScope('--line-bg'), cssScope(`--line-bg-${lineStatus}`)]);
  const clsInnerLine = clsx([cssScope(), cssScope('--line-inner'), {
    [cssScope(`--line-inner-with-${infoPosition}-info`)]: showInfo
  }]);
  const innerStyle = {
    height: strokeWidth
  };
  if (trailColor) {
    innerStyle.backgroundColor = trailColor;
  }
  return /*#__PURE__*/React.createElement(React.Fragment, null, showInfo && infoPosition === 'left' && renderSteps(), /*#__PURE__*/React.createElement("div", {
    className: clsInnerLine,
    style: innerStyle
  }, /*#__PURE__*/React.createElement("div", {
    className: clsBg,
    style: {
      width: `${percent}%`,
      [strokeColor && 'backgroundColor']: strokeColor
    }
  })), showInfo && infoPosition === 'right' && renderSteps());
};
export default Line;
