// @universe-design/react/lib/notice/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Notice = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _ErrorFilled = _interopRequireDefault(require("@universe-design/icons-react/ErrorFilled"));
var _CloseOutlined = _interopRequireDefault(require("@universe-design/icons-react/CloseOutlined"));
var _InfoFilled = _interopRequireDefault(require("@universe-design/icons-react/InfoFilled"));
var _WarningFilled = _interopRequireDefault(require("@universe-design/icons-react/WarningFilled"));
var _SucceedFilled = _interopRequireDefault(require("@universe-design/icons-react/SucceedFilled"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _reactUtilities = require("@universe-design/react-utilities");
var _button = require("../button");
var _clsx = _interopRequireDefault(require("clsx"));
var _react = _interopRequireWildcard(require("react"));
var _slot = require("../shared/slot");
var _util = require("../util");
var _hooks = require("../config-provider/hooks");
var _svgs = require("../_assets/svgs");
var _stylex2 = require("../stylex");
require("./style");
const iconMap = {
  success: _SucceedFilled.default,
  warning: _WarningFilled.default,
  error: _ErrorFilled.default,
  loading: _svgs.SpinLoadingSvg,
  info: _InfoFilled.default
};
const Notice = exports.Notice = /*#__PURE__*/_react.default.forwardRef(function (props, ref) {
  const {
    className,
    closable = false,
    align = 'left',
    type = 'info',
    theme = 'filled',
    message,
    description,
    action,
    visible: visibleProp,
    banner = false,
    onClose,
    closeText,
    closeButton,
    afterClose,
    showIcon = true,
    role = 'alert',
    icon,
    prefixCls,
    maxLine,
    ...restProps
  } = props;
  process.env.NODE_ENV !== "production" ? (0, _util.warning)(!afterClose, (0, _util.devWarning)(Notice.displayName || Notice.name, 'Close motion is not support, so afterClose is not provided.')) : void 0;
  const cssScope = (0, _hooks.useCssScope)('notice', prefixCls);
  const [visible, setVisible] = (0, _reactUtilities.useControlledState)(true, visibleProp);
  const descRef = (0, _react.useRef)(null);
  const isNoTitle = message && !description;
  let iconNode = null;
  if (showIcon) {
    if (icon) {
      iconNode = /*#__PURE__*/_react.default.createElement("div", {
        className: cssScope('__statusIcon')
      }, icon);
    }
    if (!iconNode) {
      const StatusIcon = iconMap[type] || iconMap.info;
      iconNode = /*#__PURE__*/_react.default.createElement(StatusIcon, {
        className: cssScope('__statusIcon')
      });
    }
  }
  const isClosable = closeText ? true : closable;
  const handleClick = (0, _react.useCallback)(e => {
    setVisible(false);
    onClose && onClose(e);
  }, [setVisible, onClose]);
  const closeBtn = _react.default.useMemo(() => {
    if (closeText) {
      return /*#__PURE__*/_react.default.createElement("span", {
        className: cssScope('__close'),
        onClick: handleClick
      }, closeText);
    }
    const closeButtonSlot = (0, _slot.resolveSlot)(closeButton, {
      elementType: _button.Button,
      defaultProps: {
        icon: /*#__PURE__*/_react.default.createElement(_CloseOutlined.default, null),
        size: 'small',
        type: 'icon'
      },
      renderByDefault: true
    });
    if (closeButtonSlot) {
      closeButtonSlot.props.className = (0, _clsx.default)(cssScope('__close'), closeButtonSlot.props.className);
      const originalOnClick = closeButtonSlot.props.onClick;
      closeButtonSlot.props.onClick = e => {
        originalOnClick === null || originalOnClick === void 0 ? void 0 : originalOnClick(e);
        handleClick(e);
      };
    }
    return (0, _slot.renderSlotNode)(closeButtonSlot);
  }, [closeButton, closeText, cssScope, handleClick]);
  const innerType = type === 'loading' ? 'info' : type;
  const contentStyle = maxLine ? {
    WebkitLineClamp: maxLine
  } : undefined;
  return visible ? /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
    ref: ref,
    className: (0, _clsx.default)(cssScope(), cssScope(`-${innerType}`), align === 'center' && isNoTitle && cssScope('-align-center'), closable && cssScope('-closable'), banner && cssScope('-banner'), iconNode && cssScope('-showIcon'), action && cssScope('-has-action'), message && description && cssScope('-has-title'), cssScope(`--${theme}`), className),
    role: role
  }, restProps), iconNode, /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(cssScope('__main'), isNoTitle && cssScope('__main--no-title'))
  }, message && description && /*#__PURE__*/_react.default.createElement("div", {
    className: cssScope('__title')
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(cssScope('__title__content'), (0, _stylex.default)(_stylex2.cssTypographyHeadline))
  }, message), isClosable && closeBtn), (description || isNoTitle) && /*#__PURE__*/_react.default.createElement("div", {
    ref: descRef,
    className: cssScope('__description')
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: cssScope('__description-content'),
    style: contentStyle
  }, description || message), action && /*#__PURE__*/_react.default.createElement("div", {
    className: cssScope('__description-action')
  }, action)), isClosable && isNoTitle && closeBtn)) : null;
});
