// @universe-design/react/src/empty/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import CBasic from '../demo/basic';
import CCustomize from '../demo/customize';
import CNegative from '../demo/negative';
import CNeutral from '../demo/neutral';
import CPositive from '../demo/positive';
if (!(CBasic as Story).storyName) {
  (CBasic as Story).storyName = '0. 基础用法';
  (CBasic as Story).parameters = {
    docs: {
      description: {
        story: '简单的用法。\n\n',
      },
    },
  };
}
if (!(CCustomize as Story).storyName) {
  (CCustomize as Story).storyName = '2. 自定义图片';
  (CCustomize as Story).parameters = {
    docs: {
      description: {
        story: '通过`image`自定义图标，通过`imageStyle`自定义图片样式。\n\n',
      },
    },
  };
}
if (!(CNegative as Story).storyName) {
  (CNegative as Story).storyName = '2. 负反馈空状态';
  (CNegative as Story).parameters = {
    docs: {
      description: {
        story: '负面反馈或给用户带来负面操作的示意。\n\n',
      },
    },
  };
}
if (!(CNeutral as Story).storyName) {
  (CNeutral as Story).storyName = '2. 中性反馈';
  (CNeutral as Story).parameters = {
    docs: {
      description: {
        story: '中性反馈多用于内容为空的操作反馈。\n\n',
      },
    },
  };
}
if (!(CPositive as Story).storyName) {
  (CPositive as Story).storyName = '2. 正反馈空状态';
  (CPositive as Story).parameters = {
    docs: {
      description: {
        story: '正反馈多用于正向、引导的场景。\n',
      },
    },
  };
}

export default {
  title: 'Components/Empty/官网示例',
  tags: ['autodocs'],
};

export { CBasic, CCustomize, CNegative, CNeutral, CPositive };
