// @universe-design/react/es/transfer/sx.js
//.u1es3o6n{color:@primary-content-default;}
import uue9938 from "./../.stylex/uue9938";
//.u1aocrix::before{display:block;}
import u13e45to from "./../.stylex/u13e45to";
//.u1wl9k8s::before{content:"";}
import u1hpv8lm from "./../.stylex/u1hpv8lm";
//.uf07wuq::before{box-shadow:0 0 0 .5px @line-divider-default;}
import u1pptnmf from "./../.stylex/u1pptnmf";
//.uly5x3f{width:100%;}
import u13d3rh8 from "./../.stylex/u13d3rh8";
//.u1nlzb5t{margin-left:7px;}
import u1c9ssha from "./../.stylex/u1c9ssha";
//.u1in0ah0{margin-right:7px;}
import u5tqf9 from "./../.stylex/u5tqf9";
//.ua4xqq4{height:320px;}
import u1p0qn55 from "./../.stylex/u1p0qn55";
//.u1couhl3{box-shadow:none;}
import u1vzt5dn from "./../.stylex/u1vzt5dn";
//.u1nbblvp{padding-top:0px;}
import u165peiw from "./../.stylex/u165peiw";
//.uifp7yv{padding-right:0px;}
import u24on9z from "./../.stylex/u24on9z";
//.u1ov4xf1{padding-bottom:0px;}
import uu1oa8u from "./../.stylex/uu1oa8u";
//.u1asdtw4{padding-left:0px;}
import u1pwi6nc from "./../.stylex/u1pwi6nc";
//.u1lmfglv{margin-top:0px;}
import u1kxgv9m from "./../.stylex/u1kxgv9m";
//.uibjyge{margin-right:0px;}
import unbze3l from "./../.stylex/unbze3l";
//.u1abmfm4{margin-bottom:0px;}
import u10d06an from "./../.stylex/u10d06an";
//.u9yszdx{margin-left:0px;}
import u1w6ojc6 from "./../.stylex/u1w6ojc6";
//.u17mccla{text-align:center;}
import u1ust6hi from "./../.stylex/u1ust6hi";
//.ugr6219{margin-top:auto;}
import uwejdxd from "./../.stylex/uwejdxd";
//.u1ujusj6{margin-right:auto;}
import umejf3v from "./../.stylex/umejf3v";
//.u10jk5vf{margin-bottom:auto;}
import uuhcnxi from "./../.stylex/uuhcnxi";
//.ucgxt0o{margin-left:auto;}
import uc0fbx2 from "./../.stylex/uc0fbx2";
//.u194huds{color:@text-placeholder;}
import u16q6d1w from "./../.stylex/u16q6d1w";
//.u1phki43{margin-right:8px;}
import ujgl7ix from "./../.stylex/ujgl7ix";
//.u1ugzwwg{font-size:12px;}
import u283mfx from "./../.stylex/u283mfx";
//.u71fsbu{font-weight:400;}
import us4nc6s from "./../.stylex/us4nc6s";
//.u1rvp7gu{color:@text-caption;}
import ueh72r3 from "./../.stylex/ueh72r3";
//.u1gay4b8{padding-top:15px;}
import u1hjbjne from "./../.stylex/u1hjbjne";
//.u10pi13n{position:relative;}
import u5jxv71 from "./../.stylex/u5jxv71";
//.u122n59{align-items:center;}
import ubyhozq from "./../.stylex/ubyhozq";
//.u1869bpl{justify-content:space-between;}
import uwg391m from "./../.stylex/uwg391m";
//.u1z30l8{color:@text-title;}
import u22yt5q from "./../.stylex/u22yt5q";
//.uqerorx{flex-grow:1;}
import u99gbyh from "./../.stylex/u99gbyh";
//.ur5vbr0{min-height:0px;}
import u11lnln1 from "./../.stylex/u11lnln1";
//.u22iagw{display:flex;}
import ukchsq7 from "./../.stylex/ukchsq7";
//.u6jr5hl{align-items:flex-start;}
import uqt5uie from "./../.stylex/uqt5uie";
//.u18og1rn{margin-top:15px;}
import uuf62k4 from "./../.stylex/uuf62k4";
//.u1g777gl{margin-right:15px;}
import ux5slyd from "./../.stylex/ux5slyd";
//.u19wgctu{margin-bottom:7px;}
import u557g9t from "./../.stylex/u557g9t";
//.u1abw64p{margin-left:15px;}
import u1jmgjb9 from "./../.stylex/u1jmgjb9";
//.u1khsn2v{width:calc(100% - 30px);}
import ul0aqgr from "./../.stylex/ul0aqgr";
//.u1ewtqcl{box-sizing:border-box;}
import u1p85sbs from "./../.stylex/u1p85sbs";
//.udayyex{opacity:0.5;}
import uiedlft from "./../.stylex/uiedlft";
//.u1jhi6b8{flex-grow:0;}
import ug24l6d from "./../.stylex/ug24l6d";
//.u1vk8r99{width:50%!important;}
import u120v79e from "./../.stylex/u120v79e";
//.u1b9o377{padding-top:0!important;}
import u1m79okd from "./../.stylex/u1m79okd";
//.uapxmpa{padding-bottom:0!important;}
import u1eg41r2 from "./../.stylex/u1eg41r2";
//.u1aehjj5{pointer-events:none;}
import uudhrvd from "./../.stylex/uudhrvd";
//.udcktto{margin-right:-4px;}
import uqmn7ca from "./../.stylex/uqmn7ca";
//.u1fssmi4{color:@icon-n2;}
import u1rkqst1 from "./../.stylex/u1rkqst1";
//.uycuoez{padding-left:4px;}
import u1ram9zc from "./../.stylex/u1ram9zc";
//.u8wuabp{padding-right:4px;}
import upmcqac from "./../.stylex/upmcqac";
//.u177v4lu{padding-left:8px;}
import um3isuf from "./../.stylex/um3isuf";
import stylex from '@universe-design/stylex';
import { cssTypographyBody0 } from '../stylex';
export const sxTransferLabel = {
  "props": {
    "uwmqm3": um3isuf
  }
};
export const sxSelectAll = {
  "props": {
    "uwmqm3": u1ram9zc,
    "z189sj": upmcqac
  }
};
export const sxRemoveButton = {
  "props": {
    "t21cq0": uqmn7ca,
    "sj55zd": u1rkqst1
  }
};
export const sxDisabledRemoveButton = {
  "props": {
    "Bkecrkj": uudhrvd
  }
};
export const sxPanel = {
  "props": {
    "a9b677": u120v79e,
    "z8tnut": u1m79okd,
    "Byoj8tv": u1eg41r2
  }
};
export const sxPanelHeader = {
  "props": {
    "Bh6795r": ug24l6d
  }
};
export const sxDisabledTransferCheckbox = {
  "props": {
    "abs64n": uiedlft
  }
};
export const sxSearch = {
  "props": {
    "B6of3ja": uuf62k4,
    "t21cq0": ux5slyd,
    "jrapky": u557g9t,
    "Frg6f3": u1jmgjb9,
    "a9b677": ul0aqgr,
    "B7ck84d": u1p85sbs
  }
};
export const sxPanelListWrapper = {
  "props": {
    "Bh6795r": u99gbyh,
    "sshi5w": u11lnln1,
    "mc9l5x": ukchsq7,
    "Bt984gj": uqt5uie
  }
};
export const sxSelectionSummary = {
  "props": {
    "z8tnut": u1hjbjne,
    "uwmqm3": u1ram9zc,
    "z189sj": upmcqac,
    "jrapky": u557g9t,
    "qhf8xq": u5jxv71,
    "B7ck84d": u1p85sbs,
    "mc9l5x": ukchsq7,
    "Bt984gj": ubyhozq,
    "Brf1p80": uwg391m,
    "sj55zd": u22yt5q
  }
};
export const sxListCaption = {
  "props": {
    "Be2twd7": u283mfx,
    "Bhrd7zp": us4nc6s,
    "sj55zd": ueh72r3
  }
};
export const sxClearButton = {
  "props": {
    "t21cq0": ujgl7ix
  }
};
export const sxPanelEmpty = [{
  "props": {
    "fsow6f": u1ust6hi,
    "B6of3ja": uwejdxd,
    "t21cq0": umejf3v,
    "jrapky": uuhcnxi,
    "Frg6f3": uc0fbx2,
    "sj55zd": u16q6d1w
  }
}, cssTypographyBody0];
export const sxTransfer = {
  "props": {
    "Bqenvij": u1p0qn55,
    "E5pizo": u1vzt5dn,
    "z8tnut": u165peiw,
    "z189sj": u24on9z,
    "Byoj8tv": uu1oa8u,
    "uwmqm3": u1pwi6nc,
    "B6of3ja": u1kxgv9m,
    "t21cq0": unbze3l,
    "jrapky": u10d06an,
    "Frg6f3": u1w6ojc6
  }
};
export const sxOption = {
  "props": {
    "Frg6f3": u1c9ssha,
    "t21cq0": u5tqf9
  }
};
export const sxPanelListAutoSizer = {
  "props": {
    "a9b677": u13d3rh8
  }
};
export const sxPanelListScrolling = {
  "props": {
    "Br0sdwz": u13e45to,
    "Ftih45": u1hpv8lm,
    "h8m2vh": u1pptnmf
  }
};
export const sxLabelHighlight = {
  "props": {
    "sj55zd": uue9938
  }
};
