// @universe-design/react/es/transfer/Transfer.js
import * as React from 'react';
import clsx from 'clsx';
import { ListBox } from '../shared/components/list/List';
import { TransferProvider, useTransfer } from './hooks/TransferContext';
import { useCssScope } from '../config-provider';
import { TransferSourcePanel } from './panel/SourcePanel';
import { TransferTargetPanel } from './panel/TargetPanel';
import { sxTransfer } from './sx';
export const Transfer = /*#__PURE__*/React.forwardRef((props, ref) => {
  const cssScope = useCssScope('transfer');
  const defaultChildren = React.useMemo(() => /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(TransferSourcePanel, null), /*#__PURE__*/React.createElement(TransferTargetPanel, null)), []);
  const {
    children = defaultChildren,
    style,
    className,
    ...useTransferProps
  } = props;
  const ctxValue = useTransfer(useTransferProps);
  React.useImperativeHandle(ref, () => ({
    clearSearchValue: () => ctxValue.setSearchValue('')
  }));
  return /*#__PURE__*/React.createElement(TransferProvider, {
    value: ctxValue
  }, /*#__PURE__*/React.createElement(ListBox, {
    cascaded: true,
    style: style,
    sx: [sxTransfer],
    className: clsx(cssScope(), className)
  }, children));
});
