// @universe-design/react/es/shared/components/picker/Picker.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _extends from "@babel/runtime/helpers/esm/extends";
import React, { useState, useRef, useMemo, useCallback } from 'react';
import noop from 'lodash/noop';
import clsx from 'clsx';
import TimeOutlined from '@universe-design/icons-react/TimeOutlined';
import useControlled from '../../../hooks/useControlled';
import { isFunction, getDefaultSizeAbbr, __SCREENSHOT__ } from '../../../util';
import { formatValue, parseValue } from './utils/ui';
import useSyncState from './hooks/useSyncState';
import composeRefs from './../../../util/composeRefs';
import { isEqual } from './utils/date';
import { PickerContext } from './Context';
import InputWrapper from './Comp/input-wrapper';
import { useComponentSize, useInputGroupSize, useCssScope, useLocale } from '../../../config-provider/hooks';
import useClickOutsideEffect from './hooks/useClickOutsideEffect';
import PickerTrigger from './Comp/PickerTrigger';
import PickerPanel from './Comp/PickerPanel';
import NativeInput from '../Input';
import './style';

// Wrap with class component to enable pass generic with instance method
function InnerPicker(props) {
  const timeOperationRef = React.useRef({});
  const {
    locale: localeName,
    components: {
      picker: locale
    }
  } = useLocale();
  const {
    generateConfig,
    triggerElement,
    style,
    className,
    onKeyDown = noop,
    onMouseDown = noop,
    onMouseUp = noop,
    onClick = noop,
    onClear,
    dropdownClassName,
    popupStyle,
    getPopupContainer,
    suffixIcon = /*#__PURE__*/React.createElement(TimeOutlined, null),
    allowClear,
    autoFocus,
    disabled,
    size = 'medium',
    parse,
    picker,
    mode = picker,
    placeholder = locale.timePlaceholder,
    inputReadOnly,
    format = 'HH:mm',
    error,
    onPanelChange
  } = props;
  function dateToText(date) {
    return formatValue(generateConfig, date, {
      format,
      mode: picker,
      locale,
      localeName
    });
  }
  const defaultSize = useComponentSize();
  const inputGroupSize = useInputGroupSize();
  const cmpSize = getDefaultSizeAbbr(size || inputGroupSize || defaultSize);
  const [open, setOpened] = useControlled(props, {
    defaultValuePropName: '',
    valuePropName: 'open',
    onChangePropName: 'onOpenChange',
    defaultValue: __SCREENSHOT__
  });
  const [focused, setFocused] = useState(false);

  // eslint-disable-next-line react/destructuring-assignment
  const [value, setValue] = useControlled(props);
  const [localMode, setLocalMode] = useSyncState(mode);

  /**
   * 当前视图可见的有效date
   */
  const [viewValue, setViewValue] = useSyncState(value);
  const validText = dateToText(viewValue);

  /**
   * 输入框的值
   */
  const [text, setText] = useSyncState(validText);

  /**
   * hover相关的值
   */
  const [hoverDate, setHoverDate] = useState();
  const hoverText = dateToText(hoverDate);
  const panelDivRef = useRef(null);
  const inputDivRef = useRef(null);
  const inputRef = useRef(null);
  const triggerElementRef = useRef(null);
  useClickOutsideEffect(useMemo(() => [panelDivRef, inputDivRef, triggerElementRef], []), open, useCallback(() => setOpened(false), [setOpened]));
  const cssScope = useCssScope('picker');
  const prefixCls = cssScope();
  const rootCls = clsx(cssScope(), {
    [cssScope(`--size-${cmpSize}`)]: cmpSize,
    [cssScope('--error')]: error,
    [cssScope('--focus')]: focused
  });
  function confirm(newValue) {
    if (!isEqual(generateConfig, value, newValue)) {
      setValue(newValue);
    }

    // 暂时同步value值，不然受控模式下，可能出现问题
    setViewValue(value);
    setText(dateToText(value));
    setHoverDate(undefined);
    setOpened(false);
  }
  function handleKeyDown(e) {
    onKeyDown(e);
    if (e.key !== 'Tab' && !open) {
      setOpened(true);
      return;
    }
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    e.keyCode === 13 && confirm(viewValue);
  }
  const handleInputChange = e => {
    setHoverDate(undefined);
    const editText = e.target.value;
    setText(editText);
    const date = parseValue(generateConfig, editText, {
      format,
      parse,
      mode,
      locale,
      localeName
    });
    if (date && generateConfig.isValidate(date)
    // && !(typeof disabledDate === 'function' && disabledDate(date))
    ) {
      setViewValue(date);
    }
  };
  const handleClear = () => {
    setValue(undefined);
    // 暂时同步value值，不然受控模式下，可能出现问题
    setViewValue(value);
    setOpened(false);
    onClear && onClear();
  };
  const handleMouseUp = e => {
    if (inputRef.current) {
      inputRef.current.focus();
    }
    setOpened(true);
    onMouseUp(e);
  };
  const handleFocus = e => {
    setFocused(true);
  };
  const handleBlur = e => {
    setFocused(false);
    confirm(viewValue);
  };
  let triggerElementNode;
  if (isFunction(triggerElement)) {
    triggerElementNode = triggerElement({
      text,
      hoverText,
      open,
      value
    });
  } else {
    triggerElementNode = triggerElement;
  }
  const popupElement = /*#__PURE__*/React.createElement("div", {
    onMouseDown: e => e.preventDefault(),
    ref: panelDivRef
  }, /*#__PURE__*/React.createElement(PickerContext.Provider, {
    value: {
      mode: localMode,
      setMode: setLocalMode,
      value,
      setValue,
      setOpened
    }
  }, /*#__PURE__*/React.createElement(PickerPanel, _extends({}, props, {
    prefixCls: prefixCls,
    locale: locale,
    value: viewValue,
    operationRef: timeOperationRef,
    onPanelChange: date => {
      setViewValue(date);
      setText(formatValue(generateConfig, date, {
        format,
        mode,
        locale,
        localeName
      }));
      onPanelChange && onPanelChange(date);
    }
  }))));
  return /*#__PURE__*/React.createElement(PickerTrigger, {
    open: open,
    disabled: disabled,
    setOpened: setOpened,
    dropdownClassName: dropdownClassName,
    popupStyle: popupStyle,
    popupElement: popupElement,
    getPopupContainer: getPopupContainer
  }, (() => {
    if (triggerElementNode) {
      // 如果用户传递了非 ReactElement 这里会直接报错，导致无法正常渲染
      return /*#__PURE__*/React.cloneElement(triggerElementNode, {
        ref: composeRefs(triggerElementRef,
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        triggerElementNode.ref)
      });
    }
    return /*#__PURE__*/React.createElement("div", {
      style: style,
      ref: inputDivRef,
      className: rootCls
    }, /*#__PURE__*/React.createElement(InputWrapper, {
      className: clsx(className, `${prefixCls}-dateInput`),
      showClear: Boolean(viewValue),
      onClear: handleClear,
      prefixCls: prefixCls,
      onMouseUp: handleMouseUp,
      onMouseDown: onMouseDown,
      onClick: onClick,
      focused: focused,
      suffixIcon: suffixIcon,
      disabled: disabled,
      allowClear: allowClear
    }, /*#__PURE__*/React.createElement(NativeInput, {
      size: 12,
      className: clsx({
        [`${prefixCls}-input`]: true,
        [`${prefixCls}-input-placeholder`]: Boolean(hoverText)
      }),
      placeholder: placeholder,
      disabled: disabled,
      autoFocus: autoFocus,
      ref: inputRef,
      onKeyDown: handleKeyDown,
      onFocus: handleFocus,
      onBlur: handleBlur,
      onChange: handleInputChange,
      readOnly: inputReadOnly || isFunction(format) && !isFunction(parse),
      value: hoverText || text
    })));
  })());
}
class Picker extends React.Component {
  constructor() {
    super(...arguments);
    _defineProperty(this, "pickerRef", /*#__PURE__*/React.createRef());
    _defineProperty(this, "focus", () => {
      if (this.pickerRef.current) {
        this.pickerRef.current.focus();
      }
    });
    _defineProperty(this, "blur", () => {
      if (this.pickerRef.current) {
        this.pickerRef.current.blur();
      }
    });
  }
  render() {
    return /*#__PURE__*/React.createElement(InnerPicker, _extends({}, this.props, {
      pickerRef: this.pickerRef
    }));
  }
}
export default Picker;
