// @universe-design/styles/cjs/lingo/index.dark.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.lingotoken_tag_bg_blue_pressed = exports.lingotoken_tag_bg_blue_normal = exports.lingotoken_tag_bg_blue_hover = exports.lingotoken_relationmap_lines_secondary = exports.lingotoken_relationmap_lines_primary = exports.lingotoken_relationmap_bg = exports.lingotoken_collapse_bg = exports.lingotoken_card_bg = exports.lingo_picture_toolbar_bg_hover = exports.lingo_picture_toolbar_bg = void 0;
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
exports.lingo_picture_toolbar_bg = 'rgba(41, 41, 41, 0.7)';
exports.lingo_picture_toolbar_bg_hover = 'rgba(235, 235, 235, 0.2)';
exports.lingotoken_card_bg = '#292929';
exports.lingotoken_collapse_bg = 'linear-gradient(90deg, #292929 0%, #292929 100%)';
exports.lingotoken_relationmap_bg = '#1b1c2c';
exports.lingotoken_relationmap_lines_primary = 'linear-gradient(0deg, #5e68e8 100%, #5e68e8 50%, #2a2d69 0%)';
exports.lingotoken_relationmap_lines_secondary = 'linear-gradient(0deg, #505050 100%, #505050 50%, #505050 0%)';
exports.lingotoken_tag_bg_blue_hover = 'rgba(76, 136, 255, 0.15)';
exports.lingotoken_tag_bg_blue_normal = 'rgba(76, 136, 255, 0.12)';
exports.lingotoken_tag_bg_blue_pressed = 'rgba(76, 136, 255, 0.2)';
