// @universe-design/styles/cjs/lingo/index.light.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.lingotoken_tag_bg_blue_pressed = exports.lingotoken_tag_bg_blue_normal = exports.lingotoken_tag_bg_blue_hover = exports.lingotoken_relationmap_lines_secondary = exports.lingotoken_relationmap_lines_primary = exports.lingotoken_relationmap_bg = exports.lingotoken_collapse_bg = exports.lingotoken_card_bg = exports.lingo_picture_toolbar_bg_hover = exports.lingo_picture_toolbar_bg = void 0;
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
exports.lingo_picture_toolbar_bg = 'rgba(0, 0, 0, 0.7)';
exports.lingo_picture_toolbar_bg_hover = 'rgba(255, 255, 255, 0.2)';
exports.lingotoken_card_bg = '#f8f9fa';
exports.lingotoken_collapse_bg = 'linear-gradient(90deg, #ffffff 0%, #ffffff 100%)';
exports.lingotoken_relationmap_bg = '#f8f9fe';
exports.lingotoken_relationmap_lines_primary = 'linear-gradient(0deg, #4752e6 100%, #4752e6 50%, #e9eafb 0%)';
exports.lingotoken_relationmap_lines_secondary = 'linear-gradient(0deg, #d0d3d6 100%, #d0d3d6 50%, #d0d3d6 0%)';
exports.lingotoken_tag_bg_blue_hover = 'rgba(51, 109, 244, 0.3)';
exports.lingotoken_tag_bg_blue_normal = 'rgba(51, 109, 244, 0.2)';
exports.lingotoken_tag_bg_blue_pressed = 'rgba(51, 109, 244, 0.35)';
