// @universe-design/react/lib/tag/Tag.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TagPresetColorEnum = exports.Tag = void 0;
var _u5jxv = _interopRequireDefault(require("./../.stylex/u5jxv71"));
var _u1lumzkm = _interopRequireDefault(require("./../.stylex/u1lumzkm"));
var _u11yfbjh = _interopRequireDefault(require("./../.stylex/u11yfbjh"));
var _u8o0f2y = _interopRequireDefault(require("./../.stylex/u8o0f2y"));
var _ugk9fm = _interopRequireDefault(require("./../.stylex/ugk9fm9"));
var _u7grb = _interopRequireDefault(require("./../.stylex/u7grb02"));
var _u1g1yqg = _interopRequireDefault(require("./../.stylex/u1g1yqg3"));
var _u3rhfi = _interopRequireDefault(require("./../.stylex/u3rhfi7"));
var _ux645fm = _interopRequireDefault(require("./../.stylex/ux645fm"));
var _u1km0f1u = _interopRequireDefault(require("./../.stylex/u1km0f1u"));
var _uahyb5n = _interopRequireDefault(require("./../.stylex/uahyb5n"));
var _u1agd5ph = _interopRequireDefault(require("./../.stylex/u1agd5ph"));
var _unf6cqd = _interopRequireDefault(require("./../.stylex/unf6cqd"));
var _u1c0bjt = _interopRequireDefault(require("./../.stylex/u1c0bjt1"));
var _u12o6wmt = _interopRequireDefault(require("./../.stylex/u12o6wmt"));
var _u15krktc = _interopRequireDefault(require("./../.stylex/u15krktc"));
var _u1dhzg2w = _interopRequireDefault(require("./../.stylex/u1dhzg2w"));
var _u6veevw = _interopRequireDefault(require("./../.stylex/u6veevw"));
var _uweqc5r = _interopRequireDefault(require("./../.stylex/uweqc5r"));
var _uiedlft = _interopRequireDefault(require("./../.stylex/uiedlft"));
var _uidkx = _interopRequireDefault(require("./../.stylex/uidkx17"));
var _ukrpgfj = _interopRequireDefault(require("./../.stylex/ukrpgfj"));
var _u11l27z = _interopRequireDefault(require("./../.stylex/u11l27z2"));
var _u1ei76u = _interopRequireDefault(require("./../.stylex/u1ei76u2"));
var _u1y9lr4k = _interopRequireDefault(require("./../.stylex/u1y9lr4k"));
var _uymufub = _interopRequireDefault(require("./../.stylex/uymufub"));
var _uevmt6d = _interopRequireDefault(require("./../.stylex/uevmt6d"));
var _ut0j16j = _interopRequireDefault(require("./../.stylex/ut0j16j"));
var _u165peiw = _interopRequireDefault(require("./../.stylex/u165peiw"));
var _u24on9z = _interopRequireDefault(require("./../.stylex/u24on9z"));
var _uu1oa8u = _interopRequireDefault(require("./../.stylex/uu1oa8u"));
var _u1pwi6nc = _interopRequireDefault(require("./../.stylex/u1pwi6nc"));
var _ut7q8es = _interopRequireDefault(require("./../.stylex/ut7q8es"));
var _u1o367ev = _interopRequireDefault(require("./../.stylex/u1o367ev"));
var _ujixoer = _interopRequireDefault(require("./../.stylex/ujixoer"));
var _u441rjq = _interopRequireDefault(require("./../.stylex/u441rjq"));
var _um3isuf = _interopRequireDefault(require("./../.stylex/um3isuf"));
var _u19bueuo = _interopRequireDefault(require("./../.stylex/u19bueuo"));
var _ud7edr = _interopRequireDefault(require("./../.stylex/ud7edr6"));
var _u283mfx = _interopRequireDefault(require("./../.stylex/u283mfx"));
var _u1ny05fg = _interopRequireDefault(require("./../.stylex/u1ny05fg"));
var _u1hkv3ak = _interopRequireDefault(require("./../.stylex/u1hkv3ak"));
var _u1m9gb = _interopRequireDefault(require("./../.stylex/u1m9gb09"));
var _u16uuw9n = _interopRequireDefault(require("./../.stylex/u16uuw9n"));
var _ua0xsel = _interopRequireDefault(require("./../.stylex/ua0xsel"));
var _u1fdqv2r = _interopRequireDefault(require("./../.stylex/u1fdqv2r"));
var _upmcqac = _interopRequireDefault(require("./../.stylex/upmcqac"));
var _u6qvdto = _interopRequireDefault(require("./../.stylex/u6qvdto"));
var _u1ram9zc = _interopRequireDefault(require("./../.stylex/u1ram9zc"));
var _u1vjduc = _interopRequireDefault(require("./../.stylex/u1vjduc2"));
var _u1qr18xr = _interopRequireDefault(require("./../.stylex/u1qr18xr"));
var _ug0oh3b = _interopRequireDefault(require("./../.stylex/ug0oh3b"));
var _ubyhozq = _interopRequireDefault(require("./../.stylex/ubyhozq"));
var _uugkpzq = _interopRequireDefault(require("./../.stylex/uugkpzq"));
var _u18eu6n = _interopRequireDefault(require("./../.stylex/u18eu6n2"));
var _uf7fnhg = _interopRequireDefault(require("./../.stylex/uf7fnhg"));
var _u9hwb1v = _interopRequireDefault(require("./../.stylex/u9hwb1v"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _CloseBoldOutlined = _interopRequireDefault(require("@universe-design/icons-react/CloseBoldOutlined"));
var _clsx = _interopRequireDefault(require("clsx"));
var _react = _interopRequireWildcard(require("react"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _reactUtilities = require("@universe-design/react-utilities");
var _stylex2 = require("../stylex");
var _hooks = require("../config-provider/hooks");
var _util = require("../util");
var _common = require("../avatar/common");
var _useTextOverflowPopperSidecar = require("../config-provider/hooks/useTextOverflowPopperSidecar");
var _dom = require("../util/dom");
require("./style");
//.u10pi13n{position:relative;}
//.u2gz7yw::after{display:block;}
//.u1mdlcz9::after{position:absolute;}
//.u7na3ze::after{width:24px;}
//.u1shord1::after{height:24px;}
//.u13zj6fq::after{content:"";}
//.u9eqz9b::after{top:calc(50% - 12px);}
//.ua2bdqt::after{left:-2px;}
//.u1a3p1vp{overflow:hidden;}
//.u161knb0{padding-left:2px;}
//.u1n0apg3{font-size:8px;}
//.u1hzsgrp{margin-left:2px;}
//.u1hkdmbb:hover{opacity:0.5;}
//.uc5e8lb:pressed{opacity:0.5;}
//.u1k6fduh{cursor:pointer;}
//.u1hfsmdc{opacity:0.6;}
//.u15pjodv:hover{opacity:1;}
//.u9zk3pm:pressed{opacity:1;}
//.ujf1xye{margin-left:4px;}
//.udayyex{opacity:0.5;}
//.ui64zpg{flex-shrink:0;}
//.u1q9cn9r{padding-right:10.666666666666666px;}
//.u101vrxs{padding-right:6.666666666666667px;}
//.u1a5o2fe{padding-right:5.333333333333333px;}
//.u1szoe96{width:32px;}
//.uq4mcun{width:24px;}
//.u64fuq3{width:20px;}
//.ujw5fx7{width:16px;}
//.u1nbblvp{padding-top:0px;}
//.uifp7yv{padding-right:0px;}
//.u1ov4xf1{padding-bottom:0px;}
//.u1asdtw4{padding-left:0px;}
//.u4d9j23{justify-content:center;}
//.u4ybsrx{font-size:16px;}
//.u1d2rq10{height:32px;}
//.u19lj068{padding-right:8px;}
//.u177v4lu{padding-left:8px;}
//.uses1vf{font-size:14px;}
//.urvgh55{height:24px;}
//.u1ugzwwg{font-size:12px;}
//.ujamq6b{height:20px;}
//.u1sgzk6v{padding-right:6px;}
//.u1bg5zqg{padding-left:6px;}
//.unmn6fi{font-size:10px;}
//.ud461yt{height:16px;}
//.u1g0x7ka{padding-top:0;}
//.u8wuabp{padding-right:4px;}
//.u1qch9an{padding-bottom:0;}
//.uycuoez{padding-left:4px;}
//.udrzuqr{cursor:not-allowed;}
//.u113hnb5{cursor:text;}
//.utuwxu6{display:inline-flex;}
//.u122n59{align-items:center;}
//.uz5stix{white-space:nowrap;}
//.u7o03ib{transition-property:all;}
//.u15l5c9n{transition-duration:0.2s;}
//.u6dzj5z{max-width:100%;}
// xs
const tagSizeConfigXsFontSize = 10;
const tagSizeConfigXsCloseIconSize = 8;
const tagSizeConfigXsHeight = 16;
const tagSizeConfigXsPaddingX = 4;

// sm
const tagSizeConfigSmFontSize = 12;
const tagSizeConfigSmCloseIconSize = 10;
const tagSizeConfigSmHeight = 20;
const tagSizeConfigSmPaddingX = 6;

// md
const tagSizeConfigMdFontSize = 14;
const tagSizeConfigMdCloseIconSize = 12;
const tagSizeConfigMdHeight = 24;
const tagSizeConfigMdPaddingX = 6;

// lg
const tagSizeConfigLgFontSize = 16;
const tagSizeConfigLgCloseIconSize = 14;
const tagSizeConfigLgHeight = 32;
const tagSizeConfigLgPaddingX = 8;
const sxTag = {
  root: {
    base: _stylex.default.compose(_stylex2.cssComponentBase, _stylex2.sxBorderRadius.xs, {
      "props": {
        "Bceei9c": _u1qr18xr.default,
        "mc9l5x": _ug0oh3b.default,
        "Bt984gj": _ubyhozq.default,
        "Huce71": _uugkpzq.default,
        "Bmy1vo4": _u18eu6n.default,
        "B3o57yi": _uf7fnhg.default,
        "B2u0y6b": _u9hwb1v.default
      }
    }, _stylex2.sxTransitionTimingFunction.standard),
    disabled: {
      "props": {
        "Bceei9c": _u1vjduc.default
      }
    },
    xs: {
      "props": {
        "Be2twd7": _u16uuw9n.default,
        "Bqenvij": _ua0xsel.default,
        "z8tnut": _u1fdqv2r.default,
        "z189sj": _upmcqac.default,
        "Byoj8tv": _u6qvdto.default,
        "uwmqm3": _u1ram9zc.default
      }
    },
    sm: {
      "props": {
        "Be2twd7": _u283mfx.default,
        "Bqenvij": _u1ny05fg.default,
        "z8tnut": _u1fdqv2r.default,
        "z189sj": _u1hkv3ak.default,
        "Byoj8tv": _u6qvdto.default,
        "uwmqm3": _u1m9gb.default
      }
    },
    md: {
      "props": {
        "Be2twd7": _u19bueuo.default,
        "Bqenvij": _ud7edr.default,
        "z8tnut": _u1fdqv2r.default,
        "z189sj": _u1hkv3ak.default,
        "Byoj8tv": _u6qvdto.default,
        "uwmqm3": _u1m9gb.default
      }
    },
    lg: {
      "props": {
        "Be2twd7": _u1o367ev.default,
        "Bqenvij": _ujixoer.default,
        "z8tnut": _u1fdqv2r.default,
        "z189sj": _u441rjq.default,
        "Byoj8tv": _u6qvdto.default,
        "uwmqm3": _um3isuf.default
      }
    },
    square: {
      base: {
        "props": {
          "z8tnut": _u165peiw.default,
          "z189sj": _u24on9z.default,
          "Byoj8tv": _uu1oa8u.default,
          "uwmqm3": _u1pwi6nc.default,
          "Brf1p80": _ut7q8es.default
        }
      },
      xs: {
        "props": {
          "a9b677": _ut0j16j.default
        }
      },
      sm: {
        "props": {
          "a9b677": _uevmt6d.default
        }
      },
      md: {
        "props": {
          "a9b677": _uymufub.default
        }
      },
      lg: {
        "props": {
          "a9b677": _u1y9lr4k.default
        }
      }
    },
    round: {
      base: _stylex.default.compose({
        "props": {
          "z8tnut": _u1fdqv2r.default,
          "z189sj": _u441rjq.default,
          "Byoj8tv": _u6qvdto.default,
          "uwmqm3": _um3isuf.default
        }
      }, _stylex2.sxBorderRadius.xl),
      xs: {
        "props": {
          "z189sj": _u1ei76u.default
        }
      },
      sm: {
        "props": {
          "z189sj": _u11l27z.default
        }
      },
      md: {
        "props": {
          "z189sj": _u441rjq.default
        }
      },
      lg: {
        "props": {
          "z189sj": _ukrpgfj.default
        }
      }
    }
  },
  statusIcon: {
    base: {
      "props": {
        "Bnnss6s": _uidkx.default
      }
    },
    xs: {
      "props": {
        "Be2twd7": _u283mfx.default
      }
    }
  },
  content: {
    base: _stylex.default.compose(_stylex2.cssBlockBase, _stylex2.sxOverflowEllipsis),
    disabled: {
      "props": {
        "abs64n": _uiedlft.default
      }
    },
    xs: _stylex2.cssTypographyCaption1Hack,
    sm: _stylex2.cssTypographyBody2,
    md: _stylex2.cssTypographyBody0,
    lg: _stylex2.cssTypographyTitle5,
    status: {
      xs: _stylex2.cssTypographyCaption1Hack,
      sm: _stylex2.cssTypographyCaption0,
      md: _stylex2.cssTypographyHeadline,
      lg: _stylex2.cssTypographyTitle4
    },
    hasIcon: {
      "props": {
        "Frg6f3": _uweqc5r.default
      }
    }
  },
  closeIcon: {
    base: _stylex.default.compose({
      "props": {
        "Bceei9c": _u12o6wmt.default,
        "Frg6f3": _uweqc5r.default,
        "abs64n": _u15krktc.default,
        "Bmy1vo4": _u18eu6n.default,
        "B3o57yi": _uf7fnhg.default,
        "B7zu5sd": _u1dhzg2w.default,
        "Bthtq7b": _u6veevw.default
      }
    }, _stylex2.sxTransitionTimingFunction.standard),
    disabled: {
      "props": {
        "abs64n": _uiedlft.default,
        "B7zu5sd": _unf6cqd.default,
        "Bthtq7b": _u1c0bjt.default
      }
    },
    xs: {
      "props": {
        "Be2twd7": _uahyb5n.default,
        "Frg6f3": _u1agd5ph.default
      }
    },
    sm: {
      "props": {
        "Be2twd7": _u16uuw9n.default
      }
    },
    md: {
      "props": {
        "Be2twd7": _u283mfx.default
      }
    },
    lg: {
      "props": {
        "Be2twd7": _u19bueuo.default
      }
    }
  }
};
function useTagCssScope() {
  return (0, _hooks.useCssScope)('tag');
}

/**
 * 废弃的 color 命名，以及新命名
 */
const CompatibleColorMap = {
  'neutral-select': 'neutral-option',
  'neutral-light': 'neutral-solid'
};
const TagPresetColorStatusEnum = [
// Status Tag
'waiting', 'success', 'info', 'warning', 'error', 'missing'];
const TagPresetColorEnum = exports.TagPresetColorEnum = [
// Neutral
'neutral', 'neutral-solid', 'neutral-option', 'neutral-ghost', 'neutral-ghost-transparent',
// Attribute Tag
'blue', 'wathet', 'turquoise', 'green', 'lime', 'yellow', 'orange', 'red', 'violet', 'purple', 'indigo', 'carmine',
// Solid Attribute Tag
'blue-solid', 'wathet-solid', 'turquoise-solid', 'green-solid', 'lime-solid', 'yellow-solid', 'orange-solid', 'red-solid', 'violet-solid', 'purple-solid', 'indigo-solid', 'carmine-solid',
// Option Tag
'blue-option', 'wathet-option', 'turquoise-option', 'green-option', 'lime-option', 'yellow-option', 'orange-option', 'red-option', 'violet-option', 'purple-option', 'indigo-option', 'carmine-option',
// Ghost Tag
'blue-ghost', 'carmine-ghost', 'red-ghost', ...TagPresetColorStatusEnum,
// Avatar Tag
'neutral-avatar', 'blue-avatar', 'orange-avatar'];

/**
 * @public
 */

const getPresetColor = color => {
  const compatibleColor = CompatibleColorMap[color];
  process.env.NODE_ENV !== "production" ? (0, _util.warning)(Boolean(!compatibleColor), (0, _util.devWarning)('Tag', `color value \`${color}\` is deprecated, please use \`${compatibleColor}\` instead.`)) : void 0;
  const _color = compatibleColor !== null && compatibleColor !== void 0 ? compatibleColor : color;
  if (TagPresetColorEnum.includes(_color)) {
    return [true, _color];
  }
  return [false, _color];
};

/**
 * @public
 */

const tagAvatarSize = {
  'extra-small': 12,
  small: 16,
  medium: 20,
  large: 28
};
function getAvatarSizeFromTagSize(size) {
  return tagAvatarSize[size];
}
const Tag = exports.Tag = /*#__PURE__*/_react.default.forwardRef((props, ref) => {
  var _icon$props;
  const {
    color: colorProp = 'neutral',
    className,
    size: sizeProp = 'medium',
    icon,
    children,
    closable = false,
    visible: visibleProp,
    onClose,
    style,
    shape = 'rect',
    disabled = false,
    sx,
    ...restProps
  } = props;
  const size = (0, _util.getSizeAbbr)(sizeProp);
  const cssScope = useTagCssScope();
  const TextOverflowPopper = (0, _useTextOverflowPopperSidecar.useTextOverflowPopperSidecar)();
  const [isPresetColor, color] = getPresetColor(colorProp);
  const tagStyle = {
    backgroundColor: color && !isPresetColor ? color : undefined,
    ...style
  };
  const [visible, setVisible] = (0, _reactUtilities.useControlledState)(true, visibleProp);
  const onCloseClick = _react.default.useCallback(e => {
    // 这个逻辑参考其他开源框架，不过感觉有受控 visible 不需要这个 prevented 判断
    e.stopPropagation();
    onClose === null || onClose === void 0 ? void 0 : onClose(e);
    if (e.defaultPrevented) {
      return;
    }
    setVisible(false);
  }, [onClose, setVisible]);
  const contentCls = (0, _clsx.default)(cssScope('__content'), icon && cssScope('__content-has-icon'));
  const contentSx = [sxTag.content.base, disabled && sxTag.content.disabled, icon && sxTag.content.hasIcon, icon && size === 'xs' && {
    "props": {
      "Frg6f3": _u1agd5ph.default
    }
  }, size && (TagPresetColorStatusEnum.includes(color) ? sxTag.content.status[size] : sxTag.content[size])];
  const contentRef = (0, _react.useRef)(null);
  /** extra-small 因 zoom 问题, 暂时关闭 Tooltip */
  const contentNode = (() => {
    // square tag 没有地方放 children 了
    if (!(0, _util.isNonNullable)(children) || shape === 'square') {
      return null;
    }
    return /*#__PURE__*/_react.default.createElement("div", {
      ref: contentRef,
      className: (0, _clsx.default)(contentCls, (0, _stylex.default)(contentSx))
    }, children);
  })();
  return visible ? /*#__PURE__*/_react.default.createElement(TextOverflowPopper, {
    title: children,
    checkOverflow: () => {
      if (contentRef.current) {
        return (0, _dom.isHorizontalOverflowed)(contentRef.current);
      }
      return false;
    }
  }, /*#__PURE__*/_react.default.createElement("span", (0, _extends2.default)({
    ref: ref,
    className: (0, _clsx.default)(cssScope(), className, size && cssScope(`-${size}`), shape && cssScope(`-${shape}`), cssScope(!isPresetColor ? '-other-color' : `-${color}`), icon && cssScope('-has-icon'), disabled && cssScope('-disabled'), (0, _stylex.default)(sxTag.root.base, disabled && sxTag.root.disabled, size && sxTag.root[size], (shape === 'round' || shape === 'square') && sxTag.root[shape].base, size && (shape === 'round' || shape === 'square') && sxTag.root[shape][size], shape === 'round' && icon && {
      "props": {
        "z8tnut": _u165peiw.default,
        "uwmqm3": _u1km0f1u.default,
        "Byoj8tv": _uu1oa8u.default
      }
    }, {
      "props": {
        "Bpg54ce": _ux645fm.default
      }
    }, sx)),
    style: tagStyle
  }, restProps), icon ? /*#__PURE__*/(0, _react.cloneElement)(icon, {
    className: (0, _clsx.default)(cssScope('__status-icon'), icon === null || icon === void 0 ? void 0 : (_icon$props = icon.props) === null || _icon$props === void 0 ? void 0 : _icon$props.className, (0, _stylex.default)(sxTag.statusIcon.base, size === 'xs' && sxTag.statusIcon[size])),
    ...((0, _common.isAvatar)(icon) ? {
      size: getAvatarSizeFromTagSize(sizeProp)
    } : {})
  }) : null, contentNode, closable && shape !== 'square' ? /*#__PURE__*/_react.default.createElement(_CloseBoldOutlined.default, {
    className: (0, _clsx.default)(cssScope('__close-icon'), (0, _stylex.default)(sxTag.closeIcon.base, disabled && sxTag.closeIcon.disabled, size && sxTag.closeIcon[size], {
      "props": {
        "qhf8xq": _u5jxv.default,
        "ap17g6": _u1lumzkm.default,
        "E3zdtr": _u11yfbjh.default,
        "a2br6o": _u8o0f2y.default,
        "Dlnsje": _ugk9fm.default,
        "Bsft5z2": _u7grb.default,
        "bn5sak": _u1g1yqg.default,
        "Eqx8gd": _u3rhfi.default
      }
    })),
    onClick: onCloseClick
  }) : null)) : null;
});
if (process.env.NODE_ENV !== "production") {
  Tag.displayName = 'Tag';
}
