// @universe-design/react/lib/upload/uploadList/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UploadTextFileList = exports.UploadSimpleTextFileList = exports.UploadPictureFileList = exports.UploadCardFileList = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _uploadFile = require("../uploadFile");
var _UploadContext = require("../hooks/UploadContext");
var _popover = require("../../popover");
var _space = require("../../space");
// UploadCardFileList

const defaultCardItemRenderer = file => /*#__PURE__*/_react.default.createElement(_uploadFile.UploadCardFile, {
  file: file
});
const UploadCardFileList = props => {
  const {
    itemRenderer = defaultCardItemRenderer,
    ...spaceProps
  } = props;
  const fileList = (0, _UploadContext.useUploadContextSelector)(ctx => ctx.fileList);
  return /*#__PURE__*/_react.default.createElement(_space.Space, (0, _extends2.default)({
    inline: false,
    direction: "horizontal"
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    ,
    size: [40, 16],
    wrap: true,
    align: "start"
  }, spaceProps, {
    itemKey: index => fileList[index].uid || index
  }), fileList.map(file => /*#__PURE__*/_react.default.createElement(_react.default.Fragment, {
    key: file.uid
  }, itemRenderer(file))));
};

// UploadTextFileList
exports.UploadCardFileList = UploadCardFileList;
const defaultTextItemRenderer = file => /*#__PURE__*/_react.default.createElement(_uploadFile.UploadTextFile, {
  file: file
});
const UploadTextFileList = props => {
  const {
    itemRenderer = defaultTextItemRenderer,
    ...spaceProps
  } = props;
  const fileList = (0, _UploadContext.useUploadContextSelector)(ctx => ctx.fileList);
  return /*#__PURE__*/_react.default.createElement(_space.Space, (0, _extends2.default)({
    direction: "vertical",
    inline: false
  }, spaceProps, {
    itemKey: index => fileList[index].uid || index
  }), fileList.map(file => /*#__PURE__*/_react.default.createElement(_react.default.Fragment, {
    key: file.uid
  }, itemRenderer(file))));
};

// UploadSimpleTextFileList
exports.UploadTextFileList = UploadTextFileList;
const defaultSimpleTextItemRenderer = file => /*#__PURE__*/_react.default.createElement(_uploadFile.UploadSimpleTextFile, {
  file: file
});
const UploadSimpleTextFileList = props => {
  const {
    itemRenderer = defaultSimpleTextItemRenderer,
    ...spaceProps
  } = props;
  const fileList = (0, _UploadContext.useUploadContextSelector)(ctx => ctx.fileList);
  return /*#__PURE__*/_react.default.createElement(_space.Space, (0, _extends2.default)({
    direction: "vertical",
    inline: false
  }, spaceProps, {
    itemKey: index => fileList[index].uid || index
  }), fileList.map(file => /*#__PURE__*/_react.default.createElement(_react.default.Fragment, {
    key: file.uid
  }, itemRenderer(file))));
};

// UploadPictureFileList
/**
 * @public
 */
exports.UploadSimpleTextFileList = UploadSimpleTextFileList;
const UploadPictureFileList = props => {
  const {
    maxDisplayedPictureCount,
    uploadTrigger,
    itemRenderer = file => /*#__PURE__*/_react.default.createElement(_uploadFile.UploadPictureFile, {
      file: file
    }),
    ...spaceProps
  } = props;
  const {
    fileList,
    rootPrefixCls
  } = (0, _UploadContext.useUploadContext)();
  let extraPictureList = [];
  let defaultDisplayedFiles = fileList;
  if (maxDisplayedPictureCount && fileList.length > maxDisplayedPictureCount) {
    extraPictureList = fileList.slice(maxDisplayedPictureCount - 1);
    defaultDisplayedFiles = fileList.slice(0, maxDisplayedPictureCount - 1);
  }
  const firstExtraFile = extraPictureList[0];
  const imgSrc = firstExtraFile ? firstExtraFile.thumbUrl || firstExtraFile.url : null;

  /**
   * space 默认会使用 index 作为 itemKey
   * 会使得 uploadTrigger 的 index 不断增加。导致 key 变化引发节点销毁，
   * 进而上传动作被取消
   */
  const itemKey = index => {
    if (index < defaultDisplayedFiles.length) {
      return String(defaultDisplayedFiles[index].uid);
    }
    if (extraPictureList.length && index === defaultDisplayedFiles.length) {
      return String(extraPictureList[0].uid);
    }
    return 'upload-trigger';
  };
  return (
    /*#__PURE__*/
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    _react.default.createElement(_space.Space, (0, _extends2.default)({
      wrap: true,
      inline: false,
      size: [12, 6],
      itemKey: itemKey
    }, spaceProps), defaultDisplayedFiles.map(file => /*#__PURE__*/_react.default.createElement(_react.default.Fragment, {
      key: file.uid
    }, itemRenderer(file))), extraPictureList.length > 0 && /*#__PURE__*/_react.default.createElement(_popover.Popover, {
      destroyOnHide: false,
      placement: "bottom",
      content: /*#__PURE__*/_react.default.createElement("div", {
        className: `${rootPrefixCls}-file-list ${rootPrefixCls}-item--picture__extra__list`
      }, extraPictureList.map(_file => /*#__PURE__*/_react.default.createElement(_react.default.Fragment, {
        key: _file.uid
      }, itemRenderer(_file))))
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: (0, _clsx.default)(`${rootPrefixCls}-item`, `${rootPrefixCls}-item--picture`, `${rootPrefixCls}-item--picture__extra`)
    }, imgSrc && /*#__PURE__*/_react.default.createElement("img", {
      src: imgSrc
    }), /*#__PURE__*/_react.default.createElement("div", {
      className: `${rootPrefixCls}-item--picture-mask`
    }), /*#__PURE__*/_react.default.createElement("div", {
      className: `${rootPrefixCls}-item--picture__extra-count`
    }, "+", extraPictureList.length))), uploadTrigger)
  );
};
exports.UploadPictureFileList = UploadPictureFileList;
