// @universe-design/react/lib/upload/uploadList/PictureUploadTrigger.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PictureUploadTrigger = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _AddOutlined = _interopRequireDefault(require("@universe-design/icons-react/AddOutlined"));
var _clsx = _interopRequireDefault(require("clsx"));
var _react = _interopRequireDefault(require("react"));
var _forwardRef = require("../../util/forwardRef");
var _UploadContext = require("../hooks/UploadContext");
var _UploadTrigger = require("../UploadTrigger");
const PictureUploadTrigger = exports.PictureUploadTrigger = (0, _forwardRef.forwardHTMLRef)((props, ref) => {
  const {
    accept,
    multiple,
    directory,
    type,
    openFileDialogOnClick,
    ...restProps
  } = props;
  const {
    disabled,
    rootPrefixCls
  } = (0, _UploadContext.useUploadContext)();
  return /*#__PURE__*/_react.default.createElement(_UploadTrigger.UploadTrigger, {
    accept: accept,
    multiple: multiple,
    directory: directory,
    type: type,
    openFileDialogOnClick: openFileDialogOnClick
  }, /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
    className: (0, _clsx.default)({
      [`${rootPrefixCls}-picture-upload-trigger`]: true,
      [`${rootPrefixCls}-picture-upload-trigger--disabled`]: disabled,
      [`${rootPrefixCls}-picture-upload-trigger--active`]: !disabled
    }),
    ref: ref
  }, restProps), /*#__PURE__*/_react.default.createElement(_AddOutlined.default, null)));
});
