// @universe-design/react-provider/lib-commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  createContextHook: true,
  createStore: true,
  createGlobalConfig: true,
  DomainEventTargetContextProvider: true,
  useDomainEventTargetContext: true,
  LocaleNameProvider: true,
  useLocaleName: true,
  setGlobalLocaleName: true,
  getGlobalLocaleName: true,
  LOCALE_NAMES: true
};
Object.defineProperty(exports, "DomainEventTargetContextProvider", {
  enumerable: true,
  get: function () {
    return _domainEventsContext.DomainEventTargetContextProvider;
  }
});
Object.defineProperty(exports, "LOCALE_NAMES", {
  enumerable: true,
  get: function () {
    return _locale_name.LOCALE_NAMES;
  }
});
Object.defineProperty(exports, "LocaleNameProvider", {
  enumerable: true,
  get: function () {
    return _locale_name.LocaleNameProvider;
  }
});
Object.defineProperty(exports, "createContextHook", {
  enumerable: true,
  get: function () {
    return _createContextHook.createContextHook;
  }
});
Object.defineProperty(exports, "createGlobalConfig", {
  enumerable: true,
  get: function () {
    return _createGlobalConfig.createGlobalConfig;
  }
});
Object.defineProperty(exports, "createStore", {
  enumerable: true,
  get: function () {
    return _createStore.createStore;
  }
});
Object.defineProperty(exports, "getGlobalLocaleName", {
  enumerable: true,
  get: function () {
    return _locale_name.getGlobalLocaleName;
  }
});
Object.defineProperty(exports, "setGlobalLocaleName", {
  enumerable: true,
  get: function () {
    return _locale_name.setGlobalLocaleName;
  }
});
Object.defineProperty(exports, "useDomainEventTargetContext", {
  enumerable: true,
  get: function () {
    return _domainEventsContext.useDomainEventTargetContext;
  }
});
Object.defineProperty(exports, "useLocaleName", {
  enumerable: true,
  get: function () {
    return _locale_name.useLocaleName;
  }
});
var _createContextHook = require("./createContextHook");
var _createStore = require("./createStore");
var _createGlobalConfig = require("./createGlobalConfig");
var _domainEventsContext = require("./domain-events-context");
var _dateAdapters = require("./date-adapters");
Object.keys(_dateAdapters).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _dateAdapters[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _dateAdapters[key];
    }
  });
});
var _themeProvider = require("./theme-provider");
Object.keys(_themeProvider).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _themeProvider[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _themeProvider[key];
    }
  });
});
var _localeProvider = require("./locale-provider");
Object.keys(_localeProvider).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _localeProvider[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _localeProvider[key];
    }
  });
});
var _stylesheetContainer = require("./stylesheet-container");
Object.keys(_stylesheetContainer).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _stylesheetContainer[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _stylesheetContainer[key];
    }
  });
});
var _popupContainer = require("./popup-container");
Object.keys(_popupContainer).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _popupContainer[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _popupContainer[key];
    }
  });
});
var _locale_name = require("./locale_name");
