// @lark/moment-timezone/es/extend.mjs
import e from "@lark/moment-timezone-polyfill";
import { getLocalTimezone as u } from "./core.mjs";
function a(t, n) { const o = t.format("YYYY-MM-DD HH:mm"); return e.tz(o, n); }
function f({ timestamp: t, timezone: n, isAllDay: o = !1 }) { if (o && n === "UTC") {
    const r = e.tz(t * 1e3, n);
    return e(r.format("YYYY-MM-DD"));
} return e.tz(t * 1e3, n); }
function s(t) { return e().tz(t).date(); }
function x(t) { const n = e().tz(); return n ? e.tz(t, n) : t; }
function m() { return e(); }
function M() { const t = m().format("kk:mm").split(":").map(Number); return t[0] * 60 + t[1]; }
function g(t) { return e(t * 1e3); }
function p(t) { return t.unix(); }
function l() { return e().tz("UTC").startOf("day").unix(); }
function D() { return m().unix(); }
function z(t) { const n = t.format("YYYY-MM-DD"); return e(n); }
function Y(t) { return e.tz(t, u()); }
function T(t, n) { return e(t, n); }
function L({ date: t, minutesDiff: n }) { const o = ~~(n / 60), r = n - o * 60; return t.clone().hours(o).minutes(r); }
export { x as change2DefaultTimezone, a as forceChangeTimezone, s as getDateInTimezone, g as getLocalMoment, Y as getLocalMomentFromDate, m as getLocalNow, M as getLocalNowMinutes, p as getLocalUnix, D as getLocalUnixNow, L as getMomentByMinutesDiff, f as getMomentByUnix, T as getMomentFromDateString, l as getUTCTodayStartUnix, z as getWeekStartDate };
