// @universe-design/react/es/divider/sx.js
//.u1g0x7ka{padding-top:0;}
import u1fdqv2r from "./../.stylex/u1fdqv2r";
//.ude2wo7{padding-right:1em;}
import u2fkmih from "./../.stylex/u2fkmih";
//.u1qch9an{padding-bottom:0;}
import u6qvdto from "./../.stylex/u6qvdto";
//.uyan362{padding-left:1em;}
import u19tdsxb from "./../.stylex/u19tdsxb";
//.u1ewtqcl{box-sizing:border-box;}
import u1p85sbs from "./../.stylex/u1p85sbs";
//.upnsfar::before{border-top-style:dashed;}
import ultm4w8 from "./../.stylex/ultm4w8";
//.u1k6lsn2::after{border-top-style:dashed;}
import ulphrlj from "./../.stylex/ulphrlj";
//.untkm7k{border-top-style:dashed;}
import u16kelhn from "./../.stylex/u16kelhn";
//.u10pi13n{position:relative;}
import u5jxv71 from "./../.stylex/u5jxv71";
//.uptdccv{top:-.06em;}
import u1jf9bby from "./../.stylex/u1jf9bby";
//.u14t3ns0{display:inline-block;}
import u8wgnjj from "./../.stylex/u8wgnjj";
//.u1sdqz7s{height:.9em;}
import uq5x3qs from "./../.stylex/uq5x3qs";
//.u1hu3pq6{margin-top:0;}
import u57ipeq from "./../.stylex/u57ipeq";
//.u1phki43{margin-right:8px;}
import ujgl7ix from "./../.stylex/ujgl7ix";
//.u19f4twv{margin-bottom:0;}
import u1o4axus from "./../.stylex/u1o4axus";
//.uf9s3yw{margin-left:8px;}
import u5msrnt from "./../.stylex/u5msrnt";
//.umrv4ls{vertical-align:middle;}
import u1uvgrag from "./../.stylex/u1uvgrag";
//.uinvdd3{border-left-width:1px;}
import uvn5txf from "./../.stylex/uvn5txf";
//.ujik90z{border-left-style:solid;}
import ur61548 from "./../.stylex/ur61548";
//.uzxpwcz{border-left-color:@line-divider-default;}
import u1az68f from "./../.stylex/u1az68f";
//.u22iagw{display:flex;}
import ukchsq7 from "./../.stylex/ukchsq7";
//.u1hgrdrl{clear:both;}
import uxrqna1 from "./../.stylex/uxrqna1";
//.uly5x3f{width:100%;}
import u13d3rh8 from "./../.stylex/u13d3rh8";
//.u11qra4b{min-width:100%;}
import u17002dn from "./../.stylex/u17002dn";
//.uyycim2{margin-top:24px;}
import u1nqr7ng from "./../.stylex/u1nqr7ng";
//.u11qmguv{margin-right:0;}
import u1purjcd from "./../.stylex/u1purjcd";
//.uvueend{margin-bottom:24px;}
import u1aas5tn from "./../.stylex/u1aas5tn";
//.u1tyq0we{margin-left:0;}
import u2o5q4q from "./../.stylex/u2o5q4q";
//.umd54rt::before{width:95%;}
import uibgean from "./../.stylex/uibgean";
//.utqlky5::after{width:5%;}
import u3a4h1v from "./../.stylex/u3a4h1v";
//.u1q25ane::before{width:5%;}
import u16hknla from "./../.stylex/u16hknla";
//.uiefn6m::after{width:95%;}
import u1o7gc4c from "./../.stylex/u1o7gc4c";
//.u1pit55b{border-top-width:0px;}
import u1r3828z from "./../.stylex/u1r3828z";
//.uz5stix{white-space:nowrap;}
import uugkpzq from "./../.stylex/uugkpzq";
//.u17mccla{text-align:center;}
import u1ust6hi from "./../.stylex/u1ust6hi";
//.u1s5rke0::before{position:relative;}
import u1e5k9l6 from "./../.stylex/u1e5k9l6";
//.u15lm7zd::before{top:50%;}
import umyyz0k from "./../.stylex/umyyz0k";
//.ulv9ixv::before{width:50%;}
import u18cctnp from "./../.stylex/u18cctnp";
//.u1fjh9ub::before{border-top-width:1px;}
import ujxjq4t from "./../.stylex/ujxjq4t";
//.u163fonl::before{border-top-style:solid;}
import u97qbzj from "./../.stylex/u97qbzj";
//.u1v9ev1j::before{border-top-color:inherit;}
import ulvenvz from "./../.stylex/ulvenvz";
//.u22404b::before{transform:translateY(50%);}
import u1i5fk3d from "./../.stylex/u1i5fk3d";
//.uyl8oag::before{box-sizing:border-box;}
import u4w7uz9 from "./../.stylex/u4w7uz9";
//.u1wl9k8s::before{content:"";}
import u1hpv8lm from "./../.stylex/u1hpv8lm";
//.ut8057j::after{position:relative;}
import u3fxpxy from "./../.stylex/u3fxpxy";
//.ummikit::after{top:50%;}
import uwrea88 from "./../.stylex/uwrea88";
//.u1slxich::after{width:50%;}
import u7kv12x from "./../.stylex/u7kv12x";
//.ua5nh5::after{border-top-width:1px;}
import udzhurv from "./../.stylex/udzhurv";
//.u8rth92::after{border-top-style:solid;}
import uzpagyr from "./../.stylex/uzpagyr";
//.uhtl3ys::after{border-top-color:inherit;}
import u1lymrsk from "./../.stylex/u1lymrsk";
//.unbjpgb::after{transform:translateY(50%);}
import u1rg8mk8 from "./../.stylex/u1rg8mk8";
//.u1gw3sf2::after{box-sizing:border-box;}
import uabpw5u from "./../.stylex/uabpw5u";
//.u13zj6fq::after{content:"";}
import u7grb02 from "./../.stylex/u7grb02";
//.u5ogflp{border-top-width:1px;}
import u1cpcpb5 from "./../.stylex/u1cpcpb5";
//.uzkkow9{border-top-style:solid;}
import us9j13g from "./../.stylex/us9j13g";
//.u1yiz13w{border-top-color:@line-divider-default;}
import uzthjd1 from "./../.stylex/uzthjd1";
import stylex from '@universe-design/stylex';
export const sxDivider = {
  "props": {
    "B4j52fo": u1cpcpb5,
    "icvyot": us9j13g,
    "g2u3we": uzthjd1
  }
};
export const sxWithText = {
  "props": {
    "B4j52fo": u1r3828z,
    "Huce71": uugkpzq,
    "fsow6f": u1ust6hi,
    "Brfgrao": u1e5k9l6,
    "Bciustq": umyyz0k,
    "Ccq8qp": u18cctnp,
    "B0qfbqy": ujxjq4t,
    "Bdkvgpv": u97qbzj,
    "Bq4z7u6": ulvenvz,
    "Hv9wc6": u1i5fk3d,
    "Bcvre1j": u4w7uz9,
    "Ftih45": u1hpv8lm,
    "E3zdtr": u3fxpxy,
    "bn5sak": uwrea88,
    "a2br6o": u7kv12x,
    "xrcqlc": udzhurv,
    "Bm2nyyq": uzpagyr,
    "B0n5ga8": u1lymrsk,
    "b1kco5": u1rg8mk8,
    "li1rpt": uabpw5u,
    "Bsft5z2": u7grb02
  }
};
export const sxLeft = {
  "props": {
    "Bciustq": umyyz0k,
    "Ccq8qp": u16hknla,
    "bn5sak": uwrea88,
    "a2br6o": u1o7gc4c
  }
};
export const sxRight = {
  "props": {
    "Bciustq": umyyz0k,
    "Ccq8qp": uibgean,
    "bn5sak": uwrea88,
    "a2br6o": u3a4h1v
  }
};
const sxHorizontal = {
  "props": {
    "mc9l5x": ukchsq7,
    "B2qlsxq": uxrqna1,
    "a9b677": u13d3rh8,
    "Bf4jedk": u17002dn,
    "B6of3ja": u1nqr7ng,
    "t21cq0": u1purjcd,
    "jrapky": u1aas5tn,
    "Frg6f3": u2o5q4q
  }
};
const sxVertical = {
  "props": {
    "qhf8xq": u5jxv71,
    "Bhzewxz": u1jf9bby,
    "mc9l5x": u8wgnjj,
    "Bqenvij": uq5x3qs,
    "B6of3ja": u57ipeq,
    "t21cq0": ujgl7ix,
    "jrapky": u1o4axus,
    "Frg6f3": u5msrnt,
    "ha4doy": u1uvgrag,
    "B4j52fo": u1r3828z,
    "ibv6hh": uvn5txf,
    "wvpqe5": ur61548,
    "zhjwy3": u1az68f
  }
};
export const sxDirection = type => {
  if (type === 'horizontal') {
    return sxHorizontal;
  } else {
    if (type === 'vertical') {
      return sxVertical;
    }
  }
};
export const sxDashed = {
  "props": {
    "icvyot": u16kelhn
  }
};
export const sxWithTextDashed = {
  "props": {
    "Bdkvgpv": ultm4w8,
    "Bm2nyyq": ulphrlj
  }
};
export const sxTitle = {
  "props": {
    "mc9l5x": u8wgnjj,
    "z8tnut": u1fdqv2r,
    "z189sj": u2fkmih,
    "Byoj8tv": u6qvdto,
    "uwmqm3": u19tdsxb,
    "B7ck84d": u1p85sbs
  }
};
