// @universe-design/react/es/upload/hooks/useUploadTrigger.js
import React from 'react';
import uid from '../utils/uid';
import { file2Obj, getFileItem, LIST_IGNORE, updateFileList } from '../utils';
import defaultRequest from '../utils/request';
import { useEventCallback } from '../../hooks';
export const useUploadTrigger = (props, ctxValue) => {
  const {
    headers,
    // multiple = false,
    withCredentials,
    customRequest,
    method,
    data,
    action,
    beforeUpload: beforeUploadProp,
    maxCount,
    onExceed,
    name = 'file'
  } = props;
  const {
    fileListRef,
    onChange
  } = ctxValue;
  const xhrRequests = React.useRef(new Map());
  const _isMounted = React.useRef(false);
  React.useEffect(() => {
    _isMounted.current = true;
    return () => {
      // this.abort();
      _isMounted.current = false;
    };
  }, []);
  const onError = useEventCallback((error, response, file) => {
    // removed
    if (!getFileItem(file, fileListRef.current)) {
      return;
    }
    const targetItem = file2Obj(file);
    targetItem.error = error;
    targetItem.response = response;
    targetItem.status = 'error';
    onChange(targetItem, updateFileList(targetItem, fileListRef.current));
  });
  const onProgress = useEventCallback((e, file) => {
    const fileItem = getFileItem(file, fileListRef.current);
    if (!fileItem) {
      return;
    }
    const targetItem = file2Obj(file);
    targetItem.status = 'uploading';
    targetItem.percent = e.percent;
    targetItem.loaded = e.loaded;
    const nextFileList = updateFileList(Object.assign(fileItem, targetItem), fileListRef.current);
    onChange(targetItem, nextFileList, e);
  });
  const beforeUpload = React.useCallback(async (file, fileListArgs) => {
    let parsedFile = file;
    if (beforeUploadProp) {
      const result = await beforeUploadProp(file, fileListArgs);
      if (result === false) {
        return false;
      }
      // Hack for LIST_IGNORE, we add additional info to remove from the list
      delete file[LIST_IGNORE];
      if (result === LIST_IGNORE) {
        Object.defineProperty(file, LIST_IGNORE, {
          value: true,
          configurable: true
        });
        return false;
      }
      if (typeof result === 'object' && result) {
        parsedFile = result;
      }
    }
    return parsedFile;
  }, [beforeUploadProp]);
  /**
   * Process file before upload. When all the file is ready, we start upload.
   */
  const processFile = React.useCallback(async (file, _fileList) => {
    let transformedFile = file;
    if (beforeUpload && typeof beforeUpload === 'function') {
      try {
        transformedFile = await beforeUpload(file, _fileList);
      } catch (e) {
        // Rejection will also trade as false
        transformedFile = false;
      }
      if (!transformedFile) {
        return {
          origin: file,
          parsedFile: null,
          action: '',
          data: null
        };
      }
    }

    // Get the latest action
    const mergedAction = typeof action === 'function' ? await action(file) : action || '';

    // Get latest data
    const mergedData = typeof data === 'function' ? await data(file) : data;
    const parsedData =
    // string type is from legacy `transformFile`.
    // Not sure if this will work since no related test case works with it
    (typeof transformedFile === 'object' || typeof transformedFile === 'string') && transformedFile ? transformedFile : file;
    const parsedFile = parsedData instanceof File ? parsedData : new File([parsedData], file.name, {
      type: file.type
    });
    const mergedParsedFile = parsedFile;
    mergedParsedFile.uid = file.uid;
    return {
      origin: file,
      data: mergedData,
      parsedFile: mergedParsedFile,
      action: mergedAction
    };
  }, [action, beforeUpload, data]);

  /**
   * 注意：
   *  onBatchStart onSuccess onError 会在上传的过程中异步触发，
   *  此处 用 useEventCallback 保证拿到最新值
   */
  const onBatchStart = useEventCallback(batchFileInfoList => {
    // Skip file which marked as `LIST_IGNORE`, these file will not add to file list
    const filteredFileInfoList = batchFileInfoList.filter(info => !info.file[LIST_IGNORE]);

    // Nothing to do since no file need upload
    if (!filteredFileInfoList.length) {
      return;
    }
    const objectFileList = filteredFileInfoList.map(info => ({
      ...file2Obj(info.file),
      status: 'waiting'
    }));

    // Concat new files with prev files
    let newFileList = [...fileListRef.current];
    objectFileList.forEach(fileObj => {
      // Replace file if exist
      newFileList = updateFileList(fileObj, newFileList);
    });
    if (maxCount && newFileList.length > maxCount) {
      onExceed === null || onExceed === void 0 ? void 0 : onExceed();
    }
    objectFileList.forEach((fileObj, index) => {
      // Repeat trigger `onChange` event for compatible
      let triggerFileObj = fileObj;
      if (!filteredFileInfoList[index].parsedFile) {
        // `beforeUpload` return false
        const {
          originFileObj
        } = fileObj;
        let clone;
        try {
          clone = new File([originFileObj], originFileObj.name, {
            type: originFileObj.type
          });
        } catch (e) {
          clone = new Blob([originFileObj], {
            type: originFileObj.type
          });
          clone.name = originFileObj.name;
          clone.lastModifiedDate = new Date();
          clone.lastModified = new Date().getTime();
        }
        clone.uid = fileObj.uid;
        triggerFileObj = clone;
      } else {
        // Inject `waiting` status
        fileObj.status = 'waiting';
      }
      onChange(triggerFileObj, newFileList);
    });
  });
  const onSuccess = useEventCallback((response, file, xhr) => {
    // removed
    if (!getFileItem(file, fileListRef.current)) {
      return;
    }
    try {
      if (typeof response === 'string') {
        response = JSON.parse(response);
      }
    } catch (e) {
      /* do nothing */
    }
    const targetItem = file2Obj(file);
    targetItem.status = 'success';
    targetItem.percent = 100;
    targetItem.response = response;
    targetItem.xhr = xhr;
    onChange(targetItem, updateFileList(targetItem, fileListRef.current));
  });
  const post = React.useCallback(_ref => {
    let {
      origin,
      parsedFile,
      data: _data,
      action: _action
    } = _ref;
    if (!_isMounted) {
      return;
    }
    const {
      uid: _uid
    } = origin;
    // onStart?.(origin);
    const aborter = (customRequest || defaultRequest)({
      action: _action,
      filename: name,
      data: _data,
      file: parsedFile,
      headers,
      withCredentials,
      method: method || 'post',
      onProgress: e => {
        onProgress(e, parsedFile);
      },
      onSuccess: (ret, xhr) => {
        onSuccess(ret, parsedFile, xhr);
        xhrRequests.current.delete(_uid);
      },
      onError: (err, ret) => {
        onError(err, ret, parsedFile);
        xhrRequests.current.delete(_uid);
      }
    });
    if (aborter) {
      xhrRequests.current.set(_uid, aborter);
    }
  }, [customRequest, headers, method, name, onError, onProgress, onSuccess, withCredentials]);
  const uploadFiles = React.useCallback(files => {
    const originFiles = [...files];
    const postFiles = originFiles.map(file => {
      // 如果有uid 不改变uid
      file.uid = file.uid || uid();
      return processFile(file, originFiles);
    });

    // Batch upload files
    Promise.all(postFiles).then(_fileList => {
      onBatchStart(_fileList.map(_ref2 => {
        let {
          origin,
          parsedFile
        } = _ref2;
        return {
          file: origin,
          parsedFile: parsedFile
        };
      }));
      _fileList.filter(file => file.parsedFile !== null).forEach(file => {
        post(file);
      });
    });
  }, [onBatchStart, post, processFile]);
  const abort = React.useCallback(file => {
    if (file) {
      const _uid = file.uid ? file.uid : file;
      const reqCache = xhrRequests.current.get(_uid);
      if (reqCache && reqCache.abort) {
        reqCache.abort();
      }
      xhrRequests.current.delete(_uid);
    } else {
      Object.keys(xhrRequests).forEach(_uid => {
        const reqCache = xhrRequests.current.get(_uid);
        if (reqCache && reqCache.abort) {
          reqCache.abort();
        }
        xhrRequests.current.delete(_uid);
      });
    }
  }, []);
  const retry = React.useCallback(file => {
    uploadFiles([file]);
  }, [uploadFiles]);
  return {
    onBatchStart,
    onError,
    onProgress,
    onSuccess,
    uploadFiles,
    abort,
    retry
  };
};
