// @universe-design/react/es/upload/hooks/useUpload.js
import * as React from 'react';
import { v4 as uuid } from 'uuid';
import { useCssScope } from '../../config-provider';
import useMergedState from '../../hooks/useMergedState';
import { removeFileItem } from '../utils';
import { useEventCallback } from '../../hooks/useEventCallback';
import { useLatest } from '../../hooks';
import { useUploadTrigger } from './useUploadTrigger';
export function useUpload(props, uploadRef) {
  const {
    fileList,
    defaultFileList,
    prefixCls,
    onChange: onChangeProp,
    disabled = false,
    maxCount,
    onExceed
  } = props;

  /**
   * ====================
   *   ClassName
   * ====================
   */
  const rootPrefixCls = useCssScope('upload', prefixCls)();

  /**
   * ====================
   *   File List
   * ====================
   */
  const defaultFileListFormatted = (defaultFileList || []).map(file => ({
    ...file,
    uid: file.uid ? file.uid : uuid(),
    status: file.status ? file.status : 'success'
  }));
  const [mergedFileList, setMergedFileList] = useMergedState(defaultFileListFormatted, {
    value: fileList,
    postState: list => list !== null && list !== void 0 ? list : []
  });
  const fileListRef = useLatest(mergedFileList);

  // 为文件生成默认的uid/status
  React.useMemo(() => {
    fileList === null || fileList === void 0 ? void 0 : fileList.forEach(file => {
      if (Object.isFrozen(file)) {
        return;
      }
      if (!file.status) {
        file.status = 'success';
      }
      if (!file.uid) {
        file.uid = uuid();
      }
    });
  }, [fileList]);
  const onChange = useEventCallback((file, changedFileList, event) => {
    let cloneList = [...changedFileList];
    if (maxCount === 1) {
      cloneList = cloneList.slice(-1);
    } else if (maxCount) {
      cloneList = cloneList.slice(0, maxCount);
    }
    setMergedFileList(cloneList);
    fileListRef.current = cloneList;
    const changeInfo = {
      file: file,
      fileList: cloneList
    };
    if (event) {
      changeInfo.event = event;
    }
    onChangeProp === null || onChangeProp === void 0 ? void 0 : onChangeProp(changeInfo);
  });

  /**
   * ====================
   *   Upload Trigger
   * ====================
   */
  const {
    uploadFiles,
    abort,
    retry,
    onBatchStart,
    onError,
    onProgress,
    onSuccess
  } = useUploadTrigger(props, {
    fileListRef,
    onChange
  });

  /**
   * ====================
   *   Upload File
   * ====================
   */
  const onRemove = React.useCallback(file => {
    let currentFile;
    const removedFileList = removeFileItem(file, mergedFileList);
    if (removedFileList) {
      currentFile = {
        ...file,
        status: 'removed'
      };
      mergedFileList.forEach(item => {
        const matchKey = currentFile.uid !== undefined ? 'uid' : 'name';
        if (item[matchKey] === currentFile[matchKey] && !Object.isFrozen(item)) {
          item.status = 'removed';
        }
      });
      abort(currentFile);
      onChange(currentFile, removedFileList);
    }
  }, [abort, mergedFileList, onChange]);
  const onRetry = React.useCallback(file => {
    if (file.originFileObj) {
      // 文件已损坏
      retry(file.originFileObj);
    }
  }, [retry]);

  /**
   * ====================
   *   Upload Ref
   * ====================
   * 用于测试
   */
  React.useImperativeHandle(uploadRef, () => ({
    onBatchStart,
    onSuccess,
    onProgress,
    onError,
    fileList: mergedFileList,
    upload: {
      uploadFiles,
      abort,
      retry
    }
  }));
  return {
    onChange,
    fileList: mergedFileList,
    disabled,
    rootPrefixCls,
    maxCount,
    onExceed,
    onRemove,
    onRetry,
    fileListRef,
    uploadFiles
  };
}
