// @universe-design/react/lib/steps/Description.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _u441rjq = _interopRequireDefault(require("./../.stylex/u441rjq"));
var _u1w9zb = _interopRequireDefault(require("./../.stylex/u1w9zb55"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _clsx = _interopRequireDefault(require("clsx"));
var React = _interopRequireWildcard(require("react"));
var _stylex2 = require("../stylex");
var _text = require("../text");
var _useSteps = require("./useSteps");
//.u19lj068{padding-right:8px;}
//.uxms9c{color:@icon-n3;}
const sxDescription = {
  "props": {
    "sj55zd": _u1w9zb.default
  }
};
var _default = props => {
  var _steps$index;
  const {
    stepIndex
  } = props;
  const {
    steps,
    fixedDirection,
    cssScope,
    type,
    fixedLabelPlacement,
    showIcon,
    onStepClick,
    initial
  } = (0, _useSteps.useStepsContext)();
  const index = stepIndex - initial;
  const hasIcon = type !== 'navigation' || showIcon !== false;
  // eslint-disable-next-line @typescript-eslint/no-magic-numbers
  const iconSize = type === 'dot' ? 8 : hasIcon ? 28 : 0;
  const stemItemGap = index === 0 ? 0 : type === 'navigation' ? 24 : 8;
  const gapBetweenIconTitle = hasIcon ? 8 : 0;
  let margin = 0;
  if (fixedDirection === 'vertical') {
    margin = iconSize + gapBetweenIconTitle;
  } else {
    if (type === 'navigation' || fixedLabelPlacement === 'horizontal') {
      margin = iconSize + stemItemGap + gapBetweenIconTitle;
    } else {
      margin = 0;
    }
  }
  return /*#__PURE__*/React.createElement("div", {
    style: {
      paddingLeft: margin
    },
    onClick: () => {
      onStepClick(stepIndex);
    }
  }, /*#__PURE__*/React.createElement(_text.Text, {
    as: "div",
    variant: "body0",
    key: index,
    sx: [_stylex2.cssTypographyBody2, sxDescription, fixedLabelPlacement === 'horizontal' && {
      "props": {
        "z189sj": _u441rjq.default
      }
    }],
    className: (0, _clsx.default)(cssScope('__item-description')),
    lineClamp: fixedDirection === 'horizontal' ? 2 : undefined
  }, (_steps$index = steps[index]) === null || _steps$index === void 0 ? void 0 : _steps$index.description));
};
exports.default = _default;
