// @universe-design/react/lib/steps/steps.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _u1ltv0do = _interopRequireDefault(require("./../.stylex/u1ltv0do"));
var _react = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _util = require("../util");
var _useSteps = require("./useSteps");
var _forwardRef = require("../util/forwardRef");
var _NavigationIndicator = _interopRequireDefault(require("./NavigationIndicator"));
var _Description = _interopRequireDefault(require("./Description"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
//.u18eqifj{grid-template-columns:repeat(3, 1fr);}
const Descriptions = () => {
  const {
    steps,
    count,
    initial
  } = (0, _useSteps.useStepsContext)();
  const visibleSteps = steps.slice(0, count);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, visibleSteps.map((step, index) => /*#__PURE__*/_react.default.createElement(_Description.default, {
    stepIndex: index + initial,
    key: index
  })));
};
const StepsContent = (0, _forwardRef.forwardHTMLRef)((props, ref) => {
  const {
    children
  } = props;
  const {
    style,
    cssScope,
    fixedDirection,
    activeIndex,
    initial,
    preStatus,
    fixedLabelPlacement,
    type,
    status,
    separateDescription,
    className,
    size
  } = (0, _useSteps.useStepsContext)();
  const childrenArray = (0, _util.toFlatArray)(children);
  const childrenLength = childrenArray.length;
  const isDot = type === 'dot';
  const wrapperCls = (0, _clsx.default)(cssScope(), cssScope(`--${fixedDirection}`), className, {
    [cssScope(`--${size}`)]: size,
    [cssScope('--navigation')]: type === 'navigation',
    [cssScope('--dot')]: isDot,
    [cssScope('--vertical-icon')]: fixedDirection === 'vertical' && !isDot,
    [cssScope(`--label-${fixedLabelPlacement}`)]: fixedDirection
  }, (0, _stylex.default)({
    "props": {
      "Budl1dq": _u1ltv0do.default
    }
  }));
  return /*#__PURE__*/_react.default.createElement("div", {
    className: wrapperCls,
    ref: ref,
    style: {
      ...style,
      alignItems: fixedDirection === 'horizontal' && fixedLabelPlacement === 'horizontal' || type === 'navigation' ? 'center' : undefined,
      // columnGap: 8,
      gridTemplateColumns: fixedDirection === 'horizontal' ? `repeat(${childrenLength}, 1fr)` : '1fr'
    }
  }, childrenArray.filter(_util.isNonNullable).map((child, index) => {
    if (typeof child === 'string' || typeof child === 'number' || typeof child === 'boolean' || !('props' in child)) {
      return null;
    }
    const stepNumber = initial + index;
    const childProps = {
      key: index,
      prefixCls: cssScope(),
      stepIndex: stepNumber,
      ...child.props
    };
    if (!child.props.status) {
      if (stepNumber < activeIndex) {
        childProps.status = 'finish';
      } else if (stepNumber === activeIndex) {
        childProps.status = status;
      } else {
        childProps.status = 'wait';
      }
    }
    const clx = (0, _clsx.default)({
      [cssScope('--next-error')]: status === 'error' && index === activeIndex - 1,
      [cssScope(`--pre-${preStatus}`)]: preStatus !== null
    }, childProps.className);
    childProps.className = clx;

    // preStatus = childProps.status;

    return /*#__PURE__*/(0, _react.cloneElement)(child, childProps);
  }), separateDescription && /*#__PURE__*/_react.default.createElement(Descriptions, null), type === 'navigation' ? childrenArray.map((_, index) => /*#__PURE__*/_react.default.createElement(_NavigationIndicator.default, {
    index: index,
    key: index
  })) : null);
});
const Steps = /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
  const {
    children
  } = props;
  const stepsCtxValue = (0, _useSteps.useSteps)(props);
  return /*#__PURE__*/_react.default.createElement(_useSteps.StepsContextProvider, {
    value: stepsCtxValue
  }, /*#__PURE__*/_react.default.createElement(StepsContent, {
    ref: ref
  }, children));
});
if (process.env.NODE_ENV !== "production") {
  Steps.displayName = 'Steps';
}
var _default = exports.default = Steps;
