// @universe-design/react/lib/steps/step.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _ukchsq = _interopRequireDefault(require("./../.stylex/ukchsq7"));
var _ubyhozq = _interopRequireDefault(require("./../.stylex/ubyhozq"));
var _udg2fzc = _interopRequireDefault(require("./../.stylex/udg2fzc"));
var _react = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _DoneOutlined = _interopRequireDefault(require("@universe-design/icons-react/DoneOutlined"));
var _CloseOutlined = _interopRequireDefault(require("@universe-design/icons-react/CloseOutlined"));
var _RightBoldOutlined = _interopRequireDefault(require("@universe-design/icons-react/RightBoldOutlined"));
var _react2 = require("@universe-design/stylex/react");
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _stylex2 = require("../stylex");
var _useSteps = require("./useSteps");
var _Description = _interopRequireDefault(require("./Description"));
var _text = require("../text");
//.u22iagw{display:flex;}
//.u122n59{align-items:center;}
//.u1l02sjl{height:100%;}
/** @jsxRuntime classic */
/** @jsx jsx */
/* eslint-disable max-lines-per-function */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
/**
 * `<Step />` 组件的属性。
 * @public
 */
const Step = /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
  var _steps;
  const {
    className,
    icon,
    title,
    subTitle,
    description,
    stepIndex = 0,
    status = 'wait',
    disabled
  } = props;
  const {
    registerNode,
    separateDescription,
    fixedDirection,
    clickable,
    steps,
    cssScope,
    activeIndex,
    showIcon,
    showNavIndicator,
    type,
    count,
    initial,
    progressDot,
    onStepClick
  } = (0, _useSteps.useStepsContext)();
  const prefixCls = cssScope();
  const index = stepIndex - initial;
  _react.default.useEffect(() => {
    registerNode(stepIndex, {
      description,
      disabled,
      status
    });
  }, [description, disabled, registerNode, stepIndex, status]);
  const isLast = count - 1 === index;
  const preStatus = (_steps = steps[index - 1]) === null || _steps === void 0 ? void 0 : _steps.status;
  const wrapperCls = (0, _clsx.default)(`${prefixCls}__item`, `${prefixCls}__item--${status}`, className, {
    [`${prefixCls}__item--last`]: isLast,
    [`${prefixCls}--pre-${preStatus}`]: preStatus,
    [`${prefixCls}__item--disabled`]: disabled,
    [`${prefixCls}__item--active`]: stepIndex === activeIndex,
    [`${prefixCls}__item--with-indicator`]: type === 'navigation' && showNavIndicator,
    [`${prefixCls}__item--without-icon`]: type === 'navigation' && !showIcon,
    [`${prefixCls}__item--clickable`]: clickable && !disabled
  }, ['process', 'error'].includes(status) ? (0, _stylex.default)(_stylex2.cssTypographyHeadline) : (0, _stylex.default)(_stylex2.cssTypographyBody0));
  function renderIconNode() {
    let iconNode = stepIndex + 1;
    if (type === 'dot') {
      iconNode = (0, _react2.jsx)("span", {
        className: `${prefixCls}__icon-dot`
      });
    } else if (icon) {
      iconNode = icon;
    } else if (status === 'finish') {
      iconNode = (0, _react2.jsx)(_DoneOutlined.default, null);
    } else if (status === 'error') {
      iconNode = (0, _react2.jsx)(_CloseOutlined.default, null);
    }
    const cls = (0, _clsx.default)(`${prefixCls}__icon`, {
      [`${prefixCls}--include-icon`]: typeof iconNode !== 'number'
    });
    return (0, _react2.jsx)("span", {
      className: cls
    }, iconNode);
  }
  return (0, _react2.jsx)("div", {
    className: wrapperCls,
    ref: ref,
    sx: [separateDescription && {
      "props": {
        "mc9l5x": _ukchsq.default,
        "Bt984gj": _ubyhozq.default,
        "Bqenvij": _udg2fzc.default
      }
    }]
  }, (0, _react2.jsx)("div", {
    onClick: function () {
      onStepClick(stepIndex);
    },
    className: `${prefixCls}__item-container`
  }, (0, _react2.jsx)("div", {
    className: `${prefixCls}__item-content`
  }, !isLast && (0, _react2.jsx)("div", {
    className: `${prefixCls}__item-tail`
  }), function () {
    let iconWrapper;
    if (type === 'navigation' && !showIcon) {
      return null;
    }
    iconWrapper = (0, _react2.jsx)("div", {
      className: `${prefixCls}__item-icon`
    }, renderIconNode());
    if (typeof progressDot === 'function') {
      iconWrapper = progressDot(iconWrapper, {
        index: stepIndex,
        status,
        title,
        subTitle,
        description
      });
    }
    return iconWrapper;
  }(), (0, _react2.jsx)("div", {
    className: `${prefixCls}__item-title`
  }, (0, _react2.jsx)(_text.Text, {
    className: `${prefixCls}__item-title-inner`,
    as: "div",
    lineClamp: fixedDirection === 'horizontal' ? 2 : undefined,
    variant: "plain"
  }, title, subTitle && (0, _react2.jsx)("span", {
    className: `${prefixCls}__item-subtitle`
  }, subTitle)))), !separateDescription && description && (0, _react2.jsx)(_Description.default, {
    stepIndex: stepIndex
  }), !separateDescription && !isLast && (0, _react2.jsx)("div", {
    className: `${prefixCls}__item-padding`
  })), type === 'navigation' && !isLast && (0, _react2.jsx)("div", {
    className: `${prefixCls}__item-arrow`
  }, (0, _react2.jsx)(_RightBoldOutlined.default, null)));
});
var _default = exports.default = Step;
