// @universe-design/react/lib/steps/useSteps.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StepsContextProvider = void 0;
exports.useSteps = useSteps;
exports.useStepsContext = void 0;
var _react = _interopRequireDefault(require("react"));
var _useContextSelector = require("use-context-selector");
var _useImmer = require("use-immer");
var _configProvider = require("../config-provider");
var _reactUtilities = require("@universe-design/react-utilities");
var _util = require("../util");
function useSteps(props) {
  const {
    className,
    style,
    initial = 0,
    current = 0,
    direction = 'horizontal',
    type = 'default',
    size,
    status = 'process',
    labelPlacement = 'vertical',
    progressDot = false,
    showNavIndicator = false,
    showIcon = false,
    onChange,
    children
  } = props;
  const cssScope = (0, _configProvider.useCssScope)('steps');

  // state
  const [steps, updateSteps] = (0, _useImmer.useImmer)([]);
  const [activeIndex, setActiveIndex] = (0, _reactUtilities.useControlledState)(0, current, {
    isControlledOverride: 'current' in props,
    onChange
  });
  // 这里用是否传了 onChange 作为是否可点击的判断依据
  const clickable = typeof onChange === 'function';

  /**
   * methods
   **/
  const registerNode = _react.default.useCallback((index, stepInfo) => {
    updateSteps(draft => {
      draft[index] = stepInfo;
    });
  }, [updateSteps]);
  const onStepClick = _react.default.useCallback(stepIndex => {
    if (!clickable) {
      return;
    }
    const stepDetail = steps[stepIndex - initial];
    if (!stepDetail.disabled && activeIndex !== stepIndex) {
      setActiveIndex(stepIndex);
    }
  }, [activeIndex, initial, setActiveIndex, steps, clickable]);
  let fixedLabelPlacement = labelPlacement;
  let fixedDirection = direction;
  if (direction === 'vertical') {
    fixedLabelPlacement = 'horizontal';
  }
  if (type === 'dot' && direction === 'horizontal') {
    fixedLabelPlacement = 'vertical';
  }
  if (type === 'navigation') {
    fixedDirection = 'horizontal';
    fixedLabelPlacement = showIcon ? 'horizontal' : 'vertical';
  }

  /**
   * 是否需要将 description 分离出来，单独布局
   */

  return {
    registerNode,
    steps,
    fixedDirection,
    fixedLabelPlacement,
    preStatus: null,
    showNavIndicator,
    progressDot,
    initial,
    current,
    activeIndex,
    setActiveIndex,
    status,
    size,
    style,
    className,
    cssScope,
    type,
    clickable: typeof onChange === 'function',
    showIcon,
    separateDescription: labelPlacement === 'horizontal' || type === 'navigation',
    count: (0, _util.toFlatArray)(children).length,
    onStepClick
  };
}
const {
  Provider: StepsContextProvider,
  useContext: useStepsContext
} = (0, _reactUtilities.createContextHook)('StepsContext', {
  strict: true,
  customizedContextMethods: {
    createContext: _useContextSelector.createContext,
    useContext: _useContextSelector.useContext,
    useContextSelector: _useContextSelector.useContextSelector
  }
});
exports.useStepsContext = useStepsContext;
exports.StepsContextProvider = StepsContextProvider;
