// @universe-design/react/es/date-picker/Context.js
import { createContext, useContext, useContextSelector } from 'use-context-selector';
import { devWarning, warning } from '../util';
const DatePickerContext = createContext({
  mode: 'date',
  setMode: () => process.env.NODE_ENV !== "production" ? warning(false, devWarning('DatePicker', 'You are using the default setMode from context')) : void 0,
  setValue: () => process.env.NODE_ENV !== "production" ? warning(false, devWarning('DatePicker', 'You are using the default setValue from context')) : void 0,
  setOpened: () => process.env.NODE_ENV !== "production" ? warning(false, devWarning('DatePicker', 'You are using the default setOpened from context')) : void 0,
  showToday: true,
  showTime: false,
  showNow: false,
  needConfirm: false
});
export const useDatePickerContext = () => useContext(DatePickerContext);
export const useDatePickerContextSelector = selector => useContextSelector(DatePickerContext, selector);
export const {
  Provider: DatePickerContextProvider
} = DatePickerContext;
