// @universe-design/react/es/date-picker/DatePicker.js
import CalendarOutlined from '@universe-design/icons-react/CalendarOutlined';
import clsx from 'clsx';
import noop from 'lodash/noop';
import moment from 'moment';
import React, { useCallback, useMemo, useRef, useState } from 'react';
import { useControlledState } from '@universe-design/react-utilities';
import { useComponentSize, useCssScope, useDirection, useInputGroupSize, useLocaleReceiver } from '../config-provider/hooks';
import useControlled from '../hooks/useControlled';
import useSyncState from '../hooks/useSyncState';
import NativeInput from '../shared/components/Input';
import { ListBox } from '../shared/components/list/List';
import { devWarning, getDefaultSizeAbbr, isFunction, warning } from '../util';
import composeRefs from '../util/composeRefs';
import { DatePickerContextProvider } from './Context';
import PickerPanel, { isDateValueEqual } from './panels/picker';
import { isEqual } from './utils/date';
import { getInputSize } from './utils/ui';
import generateConfig from '../shared/components/picker/generate/moment';
import './style';
import { usePopperSidecar } from '../config-provider/hooks/usePopperSidecar';
import { SlideUpDownTransition } from '../shared/components/transition/SlideUpDownTransition';
import useClickOutsideEffect from '../shared/components/picker/hooks/useClickOutsideEffect';
import InputWrapper from '../shared/components/picker/Comp/input-wrapper';
import useFormatAndParser from './hooks/useFormatAndParser';
import { isInputDateDisabled } from './utils/disabled';
import { usePopupPortalContainer } from '@universe-design/react-provider';
// nolint-next-line: long_function
export function DatePicker(props) {
  const cssScope = useCssScope('picker');
  const prefixCls = cssScope();
  const locale = useLocaleReceiver('datePicker');
  const defaultSize = useComponentSize();
  const inputGroupSize = useInputGroupSize();
  const {
    style,
    id,
    showTime = false,
    picker = 'date',
    mode = picker,
    disabledDate: disabledDateProp,
    minDate,
    maxDate,
    className,
    allowClear,
    autoFocus,
    disabled,
    // eslint-disable-next-line no-nested-ternary
    placeholder = picker === 'date' ? locale.placeholder : mode === 'month' ? locale.monthPlaceholder : locale.yearPlaceholder,
    inputReadOnly,
    format: formatProp,
    size,
    error,
    triggerElement,
    parse,
    onKeyDown = noop,
    onMouseDown = noop,
    onMouseUp = noop,
    onClick = noop,
    suffixIcon = /*#__PURE__*/React.createElement(CalendarOutlined, null),
    renderExtraFooter,
    showToday = !showTime && picker === 'date',
    getPopupContainer,
    dropdownClassName,
    popupStyle,
    startOfWeek = 0,
    dateRender,
    onClear,
    onSelect,
    defaultPickerValue,
    defaultOpen = false,
    value: valueProp,
    defaultValue,
    onChange,
    onBlur,
    needConfirm = Boolean(showTime),
    showNow = Boolean(showTime),
    hideWeekNumbers,
    janX
  } = props;
  const disabledDate = useMemo(() => {
    if (disabledDateProp === undefined) {
      return undefined;
    }
    const cache = new Map();
    return function (date) {
      const cacheKey = date.format('YYYYMMDD');
      if (cache.has(cacheKey)) {
        return cache.get(cacheKey);
      }
      const isDisabled = disabledDateProp(date);
      cache.set(cacheKey, isDisabled);
      return isDisabled;
    };
  }, [disabledDateProp]);
  const direction = useDirection();
  const Popper = usePopperSidecar();
  const cmpSize = getDefaultSizeAbbr(size || inputGroupSize || defaultSize);
  const showTimeSelection = Boolean(showTime);
  const {
    formatValue,
    parseValue,
    format
  } = useFormatAndParser({
    mode,
    format: formatProp,
    parse,
    showTime: showTimeSelection
  });
  process.env.NODE_ENV !== "production" ? warning(!(showNow && showToday), devWarning('datePicker', 'should not set showNow and showToday at same time')) : void 0;
  const dateToText = React.useCallback(date => formatValue(date), [formatValue]);
  const [value, setValue] = useControlledState(
  // 如果 valueProp 是 moment 形式，采用 valueProp
  // 如果 valueProp 是 undefined 或其他，统一采用 undefined 形式
  moment.isMoment(defaultValue) && defaultValue.isValid() ? defaultValue : undefined, moment.isMoment(valueProp) && valueProp.isValid() ? valueProp : undefined, {
    onChange,
    isEqual: isDateValueEqual,
    isControlledOverride: 'value' in props
  });

  /**
   * 当前视图可见的有效date
   */
  const [viewValue, setViewValue] = useSyncState(value, {
    isEqual: isDateValueEqual
  });
  const validText = dateToText(viewValue);

  /**
   * 输入框的值
   */
  const [text, setText] = useSyncState(validText);

  /**
   * hover相关的值
   */
  const [hoverDate, setHoverDate] = useState();
  const hoverText = dateToText(hoverDate);
  const [focused, setFocused] = useState(false);
  const [open, setOpened] = useControlled(props, {
    defaultValuePropName: '',
    valuePropName: 'open',
    onChangePropName: 'onOpenChange',
    defaultValue: defaultOpen
  });
  const panelDivRef = React.useRef(null);
  const inputDivRef = React.useRef(null);
  const inputRef = useRef(null);
  const triggerElementRef = React.useRef(null);
  const rootCls = clsx(cssScope(), {
    [cssScope(`--size-${cmpSize}`)]: cmpSize,
    [cssScope('--error')]: error,
    [cssScope('--focus')]: focused,
    [cssScope('--disabled')]: disabled
  });
  const ctxGetPopupContainer = usePopupPortalContainer();
  const confirm = useCallback(newValue => {
    if (!isEqual(value, newValue, mode, showTimeSelection)) {
      setValue(newValue);
    }

    // 暂时同步value值，不然受控模式下，可能出现问题
    setViewValue(value);
    setText(dateToText(value));
    setHoverDate(undefined);
    setOpened(false);
  }, [value, mode, showTimeSelection, setViewValue, setText, dateToText, setHoverDate, setOpened, setValue]);
  const cancel = useCallback(() => {
    setOpened(false);
    setViewValue(value);
    setText(dateToText(value));
  }, [dateToText, setOpened, setText, setViewValue, value]);
  function handleKeyDown(e) {
    onKeyDown(e);
    if (e.key !== 'Tab' && !open) {
      setOpened(true);
      return;
    }

    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    e.keyCode === 13 && confirm(viewValue);
  }
  const handleInputChange = e => {
    setHoverDate(undefined);
    const editText = e.target.value;
    setText(editText);
    const date = parseValue(editText);
    if (date !== null && date !== void 0 && date.isValid() && !isInputDateDisabled(date, {
      minDate,
      maxDate,
      disabledDate
    })) {
      setViewValue(date);
    }
  };
  const handleClear = () => {
    setValue(undefined);
    // 暂时同步value值，不然受控模式下，可能出现问题
    setViewValue(value);
    setOpened(false);
    onClear === null || onClear === void 0 ? void 0 : onClear();
  };
  const handleMouseUp = e => {
    if (inputRef.current) {
      inputRef.current.focus();
    }
    setOpened(true);
    onMouseUp(e);
  };
  const handleFocus = e => {
    setFocused(true);
  };
  const handleBlur = e => {
    onBlur === null || onBlur === void 0 ? void 0 : onBlur(e);
    setFocused(false);
    needConfirm ? cancel() : confirm(viewValue);
  };

  /**
   * 点击弹窗外部行为：
   *    1. showTime
   *       1. confirmButton: 取消
   *       2. 非 confirmButton：提交
   *    2. !showTime
   *        1. confirmButton: 取消
   *        2. 非 confirmButton：取消
   */
  useClickOutsideEffect(useMemo(() => [panelDivRef, inputDivRef, triggerElementRef], []), open, useCallback(() => {
    !needConfirm && showTime ? confirm(viewValue) : cancel();
  }, [cancel, confirm, needConfirm, showTime, viewValue]));
  const [localMode, setLocalMode] = useSyncState(mode);
  const triggerElementEnable = Boolean(isFunction(triggerElement) || triggerElement);
  const showAction = triggerElementEnable ? ['click'] : [];
  let triggerElementNode;
  if (isFunction(triggerElement)) {
    triggerElementNode = triggerElement({
      text,
      hoverText,
      open,
      value
    });
  } else {
    triggerElementNode = triggerElement;
  }
  const handleOnSelect = useCallback(newValue => {
    onSelect === null || onSelect === void 0 ? void 0 : onSelect(newValue);
    if (!needConfirm && !showTime) {
      confirm(newValue);
    } else {
      setViewValue(newValue);
    }
  }, [onSelect, needConfirm, showTime, confirm, setViewValue]);
  return /*#__PURE__*/React.createElement(Popper, {
    destroyOnHide: true,
    popupStyle: popupStyle,
    getPopupContainer: triggerNode => {
      const _getPopupContainer = getPopupContainer !== null && getPopupContainer !== void 0 ? getPopupContainer : ctxGetPopupContainer;
      return _getPopupContainer(triggerNode);
    },
    visible: open,
    action: undefined,
    showAction: showAction,
    hideAction: [],
    popupClassName: clsx(dropdownClassName, `${prefixCls}-dropdown`),
    disabled: disabled,
    onVisibleChange: setOpened,
    popup: /*#__PURE__*/React.createElement(ListBox, {
      className: `${prefixCls}__list__box`,
      onMouseDown: e => e.preventDefault(),
      ref: panelDivRef
    }, /*#__PURE__*/React.createElement(DatePickerContextProvider, {
      value: {
        mode: localMode,
        setMode: setLocalMode,
        value,
        setValue: function setDateValue(v) {
          let autoCommit = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : !needConfirm;
          if (autoCommit) {
            confirm(typeof v === 'function' ? v(viewValue) : v);
          } else {
            setViewValue(v);
          }
        },
        showToday,
        renderExtraFooter,
        setOpened,
        showTime,
        needConfirm,
        showNow
      }
    }, /*#__PURE__*/React.createElement(PickerPanel, {
      defaultViewDate: defaultPickerValue,
      disabledDate: disabledDate,
      onDateMouseEnter: setHoverDate,
      onContentMouseLeave: () => setHoverDate(undefined),
      onSelect: handleOnSelect,
      value: viewValue,
      mode: mode,
      startOfWeek: startOfWeek,
      dateRender: dateRender,
      format: format,
      onCancel: cancel,
      onConfirm: () => confirm(viewValue),
      minDate: minDate,
      maxDate: maxDate,
      hideWeekNumbers: hideWeekNumbers,
      janX: janX
    }))),
    prefixCls: prefixCls,
    popupMotion: "slide-up",
    popupPlacement: direction === 'rtl' ? 'bottomRight' : 'bottomLeft',
    enableRestoreFocus: false,
    initialFocus: null,
    showArrow: false,
    arrowHasStroke: false,
    Transition: SlideUpDownTransition,
    enableMaxHeight: false,
    defaultMainAxisOffsetWithoutArrow: 4
  }, (() => {
    if (triggerElementNode) {
      // 如果用户传递了非 ReactElement 这里会直接报错，导致无法正常渲染
      return /*#__PURE__*/React.cloneElement(triggerElementNode, {
        ref: composeRefs(triggerElementRef, triggerElementNode.ref)
      });
    }
    return /*#__PURE__*/React.createElement("div", {
      style: style,
      ref: inputDivRef,
      className: rootCls
    }, /*#__PURE__*/React.createElement(InputWrapper, {
      className: clsx(className, `${prefixCls}-dateInput`),
      showClear: Boolean(viewValue),
      onClear: handleClear,
      prefixCls: prefixCls,
      onMouseUp: handleMouseUp,
      onMouseDown: onMouseDown,
      onClick: onClick,
      focused: focused,
      suffixIcon: suffixIcon,
      disabled: disabled,
      allowClear: allowClear
    }, /*#__PURE__*/React.createElement(NativeInput, {
      id: id,
      size: getInputSize(format, generateConfig),
      className: clsx({
        [`${prefixCls}-input`]: true,
        [`${prefixCls}-input-placeholder`]: Boolean(hoverText)
      }),
      placeholder: placeholder,
      disabled: disabled,
      autoFocus: autoFocus,
      ref: inputRef,
      onKeyDown: handleKeyDown,
      onFocus: handleFocus,
      onBlur: handleBlur,
      onChange: handleInputChange,
      readOnly: inputReadOnly || isFunction(format) && !isFunction(parse),
      value: hoverText || text
    })));
  })());
}
