// @universe-design/react/es/date-picker/index.js
import { DatePicker as OriginDatePicker } from './DatePicker';
import { RangePicker as OriginRangePicker } from './RangePicker';
const DatePickerCmp = OriginDatePicker;
DatePickerCmp.RangePicker = OriginRangePicker;

/**
 * Value export
 */
// DatePicker
export const DatePicker = DatePickerCmp;
export { isValidRangeValue } from './utils/date';
export { DatePickerContextProvider } from './Context';
export { PickerPanel as DatePickerPanel } from './panels/picker';
// RangePicker
export { RangePicker, RangePickerProvider } from './RangePicker';
export { RangePickerPanel } from './RangePicker/RangePickerPanel';
export { RangePickerFooter } from './RangePicker/RangePickerFooter';
export { useRangePickerContext } from './RangePicker/Context';

// DateCalendar
export { DateCalendarUI } from './panels/DateCalendar/DateCalendarUI';
export { DateCalendar } from './panels/DateCalendar/DateCalendar';
// Enum
export { Phase, TriggerType } from './RangePicker/types';

/**
 * Type export
 */

// for docs
