// @universe-design/react/es/carousel/transition-descriptors/fade.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import { transitionTimingFunctionStandard } from '../../shared/components/transition/timingFunctionConsts';
import { LinkedList } from '../../util/datastructure/LinkedList';
import { isSlideEqual, getSlideFromAbsoluteIndex, getAbsoluteIndexFromSlide } from './utils';
export class FadeCarouselTransitionDescriptor {
  constructor() {
    var _params$transitionTim;
    let params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    _defineProperty(this, "defaults", {
      duration: 1000
    });
    _defineProperty(this, "_transitionTimingFunction", void 0);
    this._transitionTimingFunction = (_params$transitionTim = params.transitionTimingFunction) !== null && _params$transitionTim !== void 0 ? _params$transitionTim : transitionTimingFunctionStandard;
  }
  onGotoSlicklyStart(params, manager) {
    const {
      slide,
      element
    } = params;
    const {
      state: {
        slide: curSlide,
        prevSlide
      }
    } = manager;
    if (prevSlide && isSlideEqual(slide, prevSlide) || !prevSlide && isSlideEqual(slide, curSlide)) {
      this._show(element);
    } else {
      this._hide(element);
    }
    this._setPosition(element, this._getSlidePosition(slide, prevSlide !== null && prevSlide !== void 0 ? prevSlide : curSlide, manager));
  }
  onGotoSlicklyActive(params, manager) {
    const {
      slide,
      element
    } = params;
    const {
      duration,
      state: {
        slide: curSlide
      }
    } = manager;
    element.style.willChange = 'opacity';
    element.style.transition = `opacity ${duration}ms ${this._transitionTimingFunction}`;
    if (isSlideEqual(slide, curSlide)) {
      this._show(element);
    } else {
      this._hide(element);
    }
  }
  onGotoSlicklyDone(params, manager) {
    const {
      slide,
      element
    } = params;
    const {
      state: {
        slide: curSlide
      }
    } = manager;
    element.style.willChange = '';
    element.style.transition = '';
    if (isSlideEqual(slide, curSlide)) {
      this._show(element);
    } else {
      this._hide(element);
    }
  }
  onGotoDirectly(params, manager) {
    const {
      slide,
      element
    } = params;
    const {
      state: {
        slide: curSlide
      }
    } = manager;
    if (isSlideEqual(slide, curSlide)) {
      this._show(element);
    } else {
      this._hide(element);
    }
    this._setPosition(element, this._getSlidePosition(slide, curSlide, manager));
  }
  getSlidesTOBeRendered(manager) {
    // Constant
    const cacheCount = 1;
    // State
    const {
      slideCount,
      state: {
        slide: curSlide
      }
    } = manager;
    const slides = new LinkedList();
    for (let i = -cacheCount; i <= cacheCount; i++) {
      const slide = getSlideFromAbsoluteIndex(getAbsoluteIndexFromSlide(curSlide, slideCount) + i, slideCount);
      slides.append(slide);
    }
    return slides.toArray();
  }
  _getSlidePosition(slide, curSlide, manager) {
    const {
      state: {
        windowSize
      }
    } = manager;
    return {
      x: windowSize.width / 2 - manager.getSlideSize(slide.slideIndex).width / 2,
      y: 0,
      z: 0
    };
  }
  _setPosition(element, position) {
    const {
      x,
      y,
      z
    } = position;
    element.style.transform = `translate3d(${x}px, ${y}px, ${z}px)`;
  }
  _show(element) {
    element.style.opacity = '1';
    element.style.pointerEvents = '';
  }
  _hide(element) {
    element.style.opacity = '0';
    element.style.pointerEvents = 'none';
  }
}
