// @universe-design/react/lib/expandable-text/ExpandableText.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ExpandableText = void 0;
var _u5jxv = _interopRequireDefault(require("./../.stylex/u5jxv71"));
var _u184z4yl = _interopRequireDefault(require("./../.stylex/u184z4yl"));
var _u5j3c7v = _interopRequireDefault(require("./../.stylex/u5j3c7v"));
var _u1u1p7mj = _interopRequireDefault(require("./../.stylex/u1u1p7mj"));
var _u8rtfz = _interopRequireDefault(require("./../.stylex/u8rtfz1"));
var _uktrixa = _interopRequireDefault(require("./../.stylex/uktrixa"));
var _u1w892lm = _interopRequireDefault(require("./../.stylex/u1w892lm"));
var _ukchsq = _interopRequireDefault(require("./../.stylex/ukchsq7"));
var _ut0j16j = _interopRequireDefault(require("./../.stylex/ut0j16j"));
var _u19pelrb = _interopRequireDefault(require("./../.stylex/u19pelrb"));
var _u1to0gca = _interopRequireDefault(require("./../.stylex/u1to0gca"));
var _u124hjsl = _interopRequireDefault(require("./../.stylex/u124hjsl"));
var _u1kxgv9m = _interopRequireDefault(require("./../.stylex/u1kxgv9m"));
var _unbze3l = _interopRequireDefault(require("./../.stylex/unbze3l"));
var _u10d06an = _interopRequireDefault(require("./../.stylex/u10d06an"));
var _u1w6ojc = _interopRequireDefault(require("./../.stylex/u1w6ojc6"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _react = require("@universe-design/stylex/react");
var React = _interopRequireWildcard(require("react"));
var _clamp = _interopRequireDefault(require("lodash/clamp"));
var _DownOutlined = _interopRequireDefault(require("@universe-design/icons-react/DownOutlined"));
var _useImmer = require("use-immer");
var _configProvider = require("../config-provider");
var _forwardRef = require("../util/forwardRef");
var _button = require("../button");
var _FirstLineAlign = require("../shared/components/FirstLineAlign");
var _hooks = require("../hooks");
var _reactUtilities = require("@universe-design/react-utilities");
var _Expandable = require("./Expandable");
require("./style/text-collapse.less");
//.u10pi13n{position:relative;}
//.u1xynx9j{left:0px;}
//.u10k790i{right:0px;}
//.u1euv43f{position:absolute;}
//.u5gq2j6{bottom:0px;}
//.u9c4gz4{justify-content:flex-end;}
//.ubhxue7{justify-content:flex-start;}
//.u22iagw{display:flex;}
//.ujw5fx7{width:16px;}
//.uebqbo6{-webkit-mask-image:linear-gradient(90deg, rgba(0, 0, 0, 0) 0%, rgba(0, 0, 0, 1) 100%);}
//.u11fbdcc{background-color:@bg-body;}
//.u1sil6mw{overflow-y:hidden;}
//.u1lmfglv{margin-top:0px;}
//.uibjyge{margin-right:0px;}
//.u1abmfm4{margin-bottom:0px;}
//.u9yszdx{margin-left:0px;}
/** @jsxRuntime classic */
/** @jsx jsx */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const useExpandableTextCssScope = () => (0, _configProvider.useCssScope)('expandable-text');

/**
 * @public
 * @docgen
 */

const ExpandableTextParagraph = props => {
  var _lineBottoms$lineClam;
  const {
    lineClamp: lineClampProp = 3,
    children
  } = props;
  const lineClamp = (0, _clamp.default)(lineClampProp, 1, Infinity);
  const {
    isExpanded,
    updateOverflowed
  } = (0, _Expandable.useExpandableContext)();
  const cssScope = useExpandableTextCssScope();
  const paragraphRef = (0, _reactUtilities.useResizeObserver)(() => {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    reCalcLineBottoms();
  });
  const paragraphContentRef = React.useRef(null);

  /**
   * State
   */
  const [{
    lineBottoms
  }, updateState] = (0, _useImmer.useImmer)({
    lineBottoms: [0]
  });
  const reCalcLineBottoms = React.useCallback(() => {
    const paragraph = paragraphRef.current;
    const paragraphContent = paragraphContentRef.current;
    if (!paragraph || !paragraphContent) {
      return;
    }
    // 获取真实占位的 boundary
    const {
      top,
      bottom,
      height
    } = paragraph.getBoundingClientRect();
    const scaleY = height / paragraph.offsetHeight;
    const lineRects = [...paragraphContent.getClientRects()].filter(rect => rect.height && rect.width // firefox 在子元素是 div 时会多一个 width 是 0 的 rect，我们要排除掉
    );
    // console.trace('reCalcLineBottoms', { scaleY, top, bottom, lineRects });
    updateState(draft => {
      let prevLine = {
        index: 0,
        bottom: 0
      };
      draft.lineBottoms.splice(1 + lineRects.length);
      draft.lineBottoms[0] = 0;
      for (const [i, rect] of lineRects.entries()) {
        const spacingBottom = (() => {
          // eslint-disable-next-line max-nested-callbacks
          const nextLine = lineRects.slice(i + 1).find(r => r.top > rect.top); // 获取当前 rect 后第一个不同行的 rect，如果找到则把两行间隔的一半作为分界
          if (nextLine) {
            return (nextLine.top - rect.bottom) / 2;
          }
          return bottom - rect.bottom; // 如果没有找到不同行的 rect 说明是最后一行了，这时用 wrapper 的 bottom 作为边界
        })();
        /**
         * currentLineBottom 是相对于 top 的底部坐标，相当于 height，可以直接除 scaleY 得到 layout 尺寸
         */
        const currentLineBottom = Math.round((rect.bottom - top + spacingBottom) / scaleY); // 计算出当前行的下边界
        if (currentLineBottom === prevLine.bottom) {
          continue; // 如果和上次计算出的行下边界一致则说明这次的 rect 与之前的 rect 同行，这种情况出现在 white-space: pre-line 时，换行符 \n 会产生一个同行的零宽 rect
        }
        prevLine = {
          index: prevLine.index + 1,
          bottom: currentLineBottom
        };
        draft.lineBottoms[prevLine.index] = prevLine.bottom;
        // console.log('draft.lineBottoms', JSON.stringify(draft.lineBottoms));
      }
    });
    updateOverflowed(lineRects.length > lineClamp);
  }, [lineClamp, paragraphRef, updateOverflowed, updateState]);

  // lineClamp 变化的时候可以保证 reactive
  (0, _hooks.useIsomorphicLayoutEffect)(() => {
    reCalcLineBottoms();
  }, [reCalcLineBottoms]);
  return (0, _react.jsx)("div", {
    className: cssScope('__paragraph'),
    sx: [{
      "props": {
        "B6of3ja": _u1kxgv9m.default,
        "t21cq0": _unbze3l.default,
        "jrapky": _u10d06an.default,
        "Frg6f3": _u1w6ojc.default
      }
    }, !isExpanded && {
      "props": {
        "Bmxbyg5": _u124hjsl.default
      }
    }],
    style: {
      ...(isExpanded ? {} : {
        maxHeight: (_lineBottoms$lineClam = lineBottoms[lineClamp]) !== null && _lineBottoms$lineClam !== void 0 ? _lineBottoms$lineClam : lineBottoms[lineBottoms.length - 1] // 这里不直接用 webkit-line-clamp 实现是因为 safari 最后一行的底部没有 spacing 导致 trigger 定位会偏上
      })
    }
  }, (0, _react.jsx)("div", {
    ref: paragraphRef
  }, (0, _react.jsx)("span", {
    ref: paragraphContentRef,
    className: cssScope('__paragraph__content')
  }, children)));
};
const ExpandableTextTrigger = props => {
  const {
    maskColor,
    position: triggerPosition = 'end'
  } = props;
  const cssScope = useExpandableTextCssScope();
  const direction = (0, _configProvider.useDirection)();
  const {
    isExpanded,
    shouldHideTrigger,
    toggleExpanded
  } = (0, _Expandable.useExpandableContext)();
  const locale = (0, _configProvider.useLocaleReceiver)('expandableText');
  const shouldShowMask = triggerPosition === 'end' && !isExpanded;

  /**
   * Render
   */
  const maskNode = React.useMemo(() => {
    if (shouldShowMask) {
      return (0, _react.jsx)("div", {
        className: cssScope('__trigger__mask'),
        sx: [{
          "props": {
            "mc9l5x": _ukchsq.default,
            "a9b677": _ut0j16j.default,
            "Bx710cf": _u19pelrb.default,
            "De3pzq": _u1to0gca.default
          }
        }],
        style: {
          backgroundColor: maskColor
        }
      }, (0, _react.jsx)(_FirstLineAlign.EmptyElement, null));
    }
    return null;
  }, [maskColor, cssScope, shouldShowMask]);
  const buttonNode = React.useMemo(() => (0, _react.jsx)("div", {
    sx: [{
      "props": {
        "mc9l5x": _ukchsq.default
      }
    }, shouldShowMask && {
      "props": {
        "De3pzq": _u1to0gca.default
      }
    }],
    style: {
      ...(shouldShowMask && {
        backgroundColor: maskColor
      })
    }
  }, (0, _react.jsx)(_button.Button, {
    className: cssScope('__trigger__button'),
    type: "link"
    // 这里用 `DownOutlined` 而不是 `DownBoldOutlined` 是因为这里细一点好看
    ,
    endIcon: (0, _react.jsx)(_DownOutlined.default, {
      style: {
        ...(isExpanded && {
          transform: 'rotate(180deg)'
        })
      }
    }),
    onClick: toggleExpanded
  }, isExpanded ? locale.fold : locale.show)), [shouldShowMask, maskColor, cssScope, isExpanded, toggleExpanded, locale.fold, locale.show]);
  if (shouldHideTrigger) {
    return null;
  }
  return (0, _react.jsx)("div", {
    className: cssScope('__trigger'),
    sx: [{
      "props": {
        "mc9l5x": _ukchsq.default
      }
    }, triggerPosition === 'start' && [{
      "props": {
        "Brf1p80": _u1w892lm.default
      }
    }], triggerPosition === 'end' && [isExpanded && [{
      "props": {
        "Brf1p80": _uktrixa.default
      }
    }], !isExpanded && [{
      "props": {
        "qhf8xq": _u1u1p7mj.default,
        "B5kzvoi": _u8rtfz.default
      }
    }, direction === 'ltr' ? {
      "props": {
        "j35jbq": _u5j3c7v.default
      }
    } : {
      "props": {
        "oyh7mz": _u184z4yl.default
      }
    }]]]
  }, maskNode, buttonNode);
};

/**
 * 文字收起组件
 */
const ExpandableText = exports.ExpandableText = (0, _forwardRef.forwardHTMLRef)((props, ref) => {
  const {
    children,
    preventCollapseWhenExpanded,
    lineClamp,
    triggerProps,
    ...domProps
  } = props;

  /**
   * Context
   */
  const cssScope = useExpandableTextCssScope();

  /**
   * Props
   */
  const mergedDOMProps = (0, _react.useMergeComponentProps)(domProps, {
    sx: [{
      "props": {
        "qhf8xq": _u5jxv.default
      },
      "alias": cssScope()
    }]
  });
  return (0, _react.jsx)(_Expandable.ExpandableProvider, {
    preventCollapseWhenExpanded
  }, (0, _react.jsx)("div", (0, _extends2.default)({
    ref: ref
  }, mergedDOMProps), (0, _react.jsx)(ExpandableTextParagraph, {
    lineClamp: lineClamp
  }, children), (0, _react.jsx)(ExpandableTextTrigger, triggerProps)));
});
if (process.env.NODE_ENV !== "production") {
  ExpandableText.displayName = 'ExpandableText';
}

/**
 * @public
 * @docgen
 */
