// @universe-design/react/lib/expandable-text/Expandable.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useExpandableContext = exports.ExpandableProvider = void 0;
var React = _interopRequireWildcard(require("react"));
var _useImmer = require("use-immer");
var _hooks = require("../hooks");
var _reactUtilities = require("@universe-design/react-utilities");
const useExpandable = props => {
  var _isExpandedProp$defau;
  const {
    isExpanded: isExpandedProp = {},
    preventCollapseWhenExpanded = false
  } = props;

  /**
   * State
   */
  const [{
    isOverflowed
  }, updateState] = (0, _useImmer.useImmer)({
    isOverflowed: false
  });

  /**
   * isExpanded State
   */
  const [isExpanded, setExpanded] = (0, _reactUtilities.useControlledState)((_isExpandedProp$defau = isExpandedProp.defaultValue) !== null && _isExpandedProp$defau !== void 0 ? _isExpandedProp$defau : false, isExpandedProp.value, {
    onChange: isExpandedProp.onChange
  });
  const toggleExpanded = (0, _hooks.useEventCallback)(() => {
    setExpanded(!isExpanded);
  });
  const updateOverflowed = (0, _hooks.useEventCallback)(_isOverflowed => {
    updateState(draft => {
      draft.isOverflowed = _isOverflowed;
    });
  });
  return {
    /**
     * 是否隐藏 trigger
     */
    shouldHideTrigger: isExpanded && preventCollapseWhenExpanded || !isOverflowed,
    isExpanded,
    toggleExpanded,
    updateOverflowed
  };
};
const {
  Provider: ExpandableContextProvider,
  useContext: useExpandableContext
} = (0, _reactUtilities.createContextHook)('ExpandableContext', {
  strict: true
});
exports.useExpandableContext = useExpandableContext;
const ExpandableProvider = _ref => {
  let {
    children,
    ...props
  } = _ref;
  return /*#__PURE__*/React.createElement(ExpandableContextProvider, {
    value: useExpandable(props)
  }, children);
};
exports.ExpandableProvider = ExpandableProvider;
