// @universe-design/styles/cjs/ccm-ccm-chat/index.dark.atom-raw.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ccmtoken_chart_aurora_y200_raw = exports.ccmtoken_chart_aurora_y100_raw = exports.ccmtoken_chart_aurora_w600_raw = exports.ccmtoken_chart_aurora_w500_raw = exports.ccmtoken_chart_aurora_w400_raw = exports.ccmtoken_chart_aurora_w300_raw = exports.ccmtoken_chart_aurora_w200_raw = exports.ccmtoken_chart_aurora_w100_raw = exports.ccmtoken_chart_aurora_t600_raw = exports.ccmtoken_chart_aurora_t500_raw = exports.ccmtoken_chart_aurora_t400_raw = exports.ccmtoken_chart_aurora_t300_raw = exports.ccmtoken_chart_aurora_t200_raw = exports.ccmtoken_chart_aurora_t100_raw = exports.ccmtoken_chart_aurora_r600_raw = exports.ccmtoken_chart_aurora_r500_raw = exports.ccmtoken_chart_aurora_r400_raw = exports.ccmtoken_chart_aurora_r300_raw = exports.ccmtoken_chart_aurora_r200_raw = exports.ccmtoken_chart_aurora_r100_raw = exports.ccmtoken_chart_aurora_p600_raw = exports.ccmtoken_chart_aurora_p500_raw = exports.ccmtoken_chart_aurora_p400_raw = exports.ccmtoken_chart_aurora_p300_raw = exports.ccmtoken_chart_aurora_p200_raw = exports.ccmtoken_chart_aurora_p100_raw = exports.ccmtoken_chart_aurora_o600_raw = exports.ccmtoken_chart_aurora_o500_raw = exports.ccmtoken_chart_aurora_o400_raw = exports.ccmtoken_chart_aurora_o300_raw = exports.ccmtoken_chart_aurora_o200_raw = exports.ccmtoken_chart_aurora_o100_raw = exports.ccmtoken_chart_aurora_g600_raw = exports.ccmtoken_chart_aurora_g500_raw = exports.ccmtoken_chart_aurora_g400_raw = exports.ccmtoken_chart_aurora_g300_raw = exports.ccmtoken_chart_aurora_g200_raw = exports.ccmtoken_chart_aurora_g100_raw = exports.ccmtoken_chart_aurora_c600_raw = exports.ccmtoken_chart_aurora_c500_raw = exports.ccmtoken_chart_aurora_c400_raw = exports.ccmtoken_chart_aurora_c300_raw = exports.ccmtoken_chart_aurora_c200_raw = exports.ccmtoken_chart_aurora_c100_raw = exports.ccmtoken_chart_aurora_b600_raw = exports.ccmtoken_chart_aurora_b500_raw = exports.ccmtoken_chart_aurora_b400_raw = exports.ccmtoken_chart_aurora_b300_raw = exports.ccmtoken_chart_aurora_b200_raw = exports.ccmtoken_chart_aurora_b100_raw = void 0;
exports.ccmtoken_chart_aurora_y600_raw = exports.ccmtoken_chart_aurora_y500_raw = exports.ccmtoken_chart_aurora_y400_raw = exports.ccmtoken_chart_aurora_y300_raw = void 0;
// autogen by ../../scripts/gen-biz-ts-from-biz-less.ts
exports.ccmtoken_chart_aurora_b100_raw = '189, 210, 255';
exports.ccmtoken_chart_aurora_b200_raw = '143, 180, 255';
exports.ccmtoken_chart_aurora_b300_raw = '76, 136, 255';
exports.ccmtoken_chart_aurora_b400_raw = '51, 112, 235';
exports.ccmtoken_chart_aurora_b500_raw = '38, 85, 182';
exports.ccmtoken_chart_aurora_b600_raw = '25, 66, 148';
exports.ccmtoken_chart_aurora_c100_raw = '255, 194, 229';
exports.ccmtoken_chart_aurora_c200_raw = '252, 148, 207';
exports.ccmtoken_chart_aurora_c300_raw = '237, 119, 186';
exports.ccmtoken_chart_aurora_c400_raw = '194, 74, 142';
exports.ccmtoken_chart_aurora_c500_raw = '148, 56, 108';
exports.ccmtoken_chart_aurora_c600_raw = '120, 43, 87';
exports.ccmtoken_chart_aurora_g100_raw = '153, 228, 144';
exports.ccmtoken_chart_aurora_g200_raw = '105, 204, 92';
exports.ccmtoken_chart_aurora_g300_raw = '81, 186, 67';
exports.ccmtoken_chart_aurora_g400_raw = '65, 158, 52';
exports.ccmtoken_chart_aurora_g500_raw = '47, 117, 38';
exports.ccmtoken_chart_aurora_g600_raw = '33, 81, 26';
exports.ccmtoken_chart_aurora_o100_raw = '254, 200, 139';
exports.ccmtoken_chart_aurora_o200_raw = '248, 158, 68';
exports.ccmtoken_chart_aurora_o300_raw = '243, 135, 27';
exports.ccmtoken_chart_aurora_o400_raw = '219, 112, 24';
exports.ccmtoken_chart_aurora_o500_raw = '161, 83, 23';
exports.ccmtoken_chart_aurora_o600_raw = '104, 58, 18';
exports.ccmtoken_chart_aurora_p100_raw = '219, 200, 253';
exports.ccmtoken_chart_aurora_p200_raw = '197, 163, 255';
exports.ccmtoken_chart_aurora_p300_raw = '165, 117, 250';
exports.ccmtoken_chart_aurora_p400_raw = '140, 85, 236';
exports.ccmtoken_chart_aurora_p500_raw = '108, 57, 198';
exports.ccmtoken_chart_aurora_p600_raw = '85, 41, 163';
exports.ccmtoken_chart_aurora_r100_raw = '248, 158, 155';
exports.ccmtoken_chart_aurora_r200_raw = '246, 130, 126';
exports.ccmtoken_chart_aurora_r300_raw = '240, 91, 86';
exports.ccmtoken_chart_aurora_r400_raw = '209, 70, 66';
exports.ccmtoken_chart_aurora_r500_raw = '160, 51, 49';
exports.ccmtoken_chart_aurora_r600_raw = '160, 51, 49';
exports.ccmtoken_chart_aurora_t100_raw = '111, 232, 216';
exports.ccmtoken_chart_aurora_t200_raw = '51, 214, 192';
exports.ccmtoken_chart_aurora_t300_raw = '45, 190, 171';
exports.ccmtoken_chart_aurora_t400_raw = '16, 168, 147';
exports.ccmtoken_chart_aurora_t500_raw = '15, 133, 117';
exports.ccmtoken_chart_aurora_t600_raw = '29, 78, 71';
exports.ccmtoken_chart_aurora_w100_raw = '137, 223, 254';
exports.ccmtoken_chart_aurora_w200_raw = '81, 195, 240';
exports.ccmtoken_chart_aurora_w300_raw = '37, 178, 229';
exports.ccmtoken_chart_aurora_w400_raw = '16, 153, 204';
exports.ccmtoken_chart_aurora_w500_raw = '20, 108, 145';
exports.ccmtoken_chart_aurora_w600_raw = '22, 67, 89';
exports.ccmtoken_chart_aurora_y100_raw = '255, 222, 117';
exports.ccmtoken_chart_aurora_y200_raw = '251, 203, 70';
exports.ccmtoken_chart_aurora_y300_raw = '240, 182, 34';
exports.ccmtoken_chart_aurora_y400_raw = '217, 153, 4';
exports.ccmtoken_chart_aurora_y500_raw = '173, 122, 3';
exports.ccmtoken_chart_aurora_y600_raw = '111, 74, 1';
