// @byted-service/env/lib/internal/vregion.data.js
"use strict";
// This file is generated by a script. Modifying this file by hand is NOT recommended.
// To update, run `npm run update-vregion`.
Object.defineProperty(exports, "__esModule", { value: true });
exports.VDC_TO_VREGION_MAP = exports.VREGION = exports.VDC = void 0;
exports.VDC = {
    VDC_UNKNOWN: '-',
    VDC_LF: 'lf',
    VDC_HL: 'hl',
    VDC_LQ: 'lq',
    VDC_WJ: 'wj',
    VDC_AWSNC1: 'awsnc1',
    VDC_YG: 'yg',
    VDC_ZK: 'zk',
    VDC_GL: 'gl',
    VDC_TJDT: 'tjdt',
    VDC_PD: 'pd',
    VDC_HJ: 'hj',
    VDC_PDDT: 'pddt',
    VDC_YZ: 'yz',
    VDC_ZJG: 'zjg',
    VDC_JJ: 'jj',
    VDC_YJ: 'yj',
    VDC_SINFONLINE: 'sinfonline',
    VDC_SINFONLINEA: 'sinfonlinea',
    VDC_SINFONLINEC: 'sinfonlinec',
    VDC_SINFVOLCVIKING: 'sinfvolcviking',
    VDC_AGCQ: 'agcq',
    VDC_AGGDSZ: 'aggdsz',
    VDC_AGHBWH: 'aghbwh',
    VDC_AGJSNJ: 'agjsnj',
    VDC_AGLNSY: 'aglnsy',
    VDC_AGSDQD: 'agsdqd',
    VDC_AGSXLQ: 'agsxlq',
    VDC_AGSXXA: 'agsxxa',
    VDC_AGSY: 'agsy',
    VDC_ALIEC1: 'aliec1',
    VDC_ALIEC2: 'aliec2',
    VDC_ALINC2: 'alinc2',
    VDC_ALIGDSZ: 'aligdsz',
    VDC_ALISH: 'alish',
    VDC_ALISC1: 'alisc1',
    VDC_BJLGY: 'bjlgy',
    VDC_ALINC2HL: 'alinc2hl',
    VDC_ALINC2LF: 'alinc2lf',
    VDC_ALINC2LQ: 'alinc2lq',
    VDC_LFZG: 'lfzg',
    VDC_HLZG: 'hlzg',
    VDC_HJZG: 'hjzg',
    VDC_QDTOB: 'qdtob',
    VDC_QDTOBIAAS: 'qdtobiaas',
    VDC_SYKA2LARK: 'syka2lark',
    VDC_SYKALARK: 'sykalark',
    VDC_SYKA3LARK: 'syka3lark',
    VDC_LFTOBIAAS: 'lftobiaas',
    VDC_HLTOBIAAS: 'hltobiaas',
    VDC_GZTOBIAAS: 'gztobiaas',
    VDC_NTTOBIAAS: 'nttobiaas',
    VDC_ZJJHTOBDDOS: 'zjjhtobddos',
    VDC_HBSJZTOBDDOS: 'hbsjztobddos',
    VDC_SHTOBIAAS: 'shtobiaas',
    VDC_JSYZTOBDDOS: 'jsyztobddos',
    VDC_BOEXFL: 'boexfl',
    VDC_GDDGTOBDDOS: 'gddgtobddos',
    VDC_XFLONLINE: 'xflonline',
    VDC_BOETOBIAAS: 'boetobiaas',
    VDC_LFTOBFINANCE: 'lftobfinance',
    VDC_SHTOBFINANCE: 'shtobfinance',
    VDC_ZJTZTOBDDOS: 'zjtztobddos',
    VDC_BJTOBIAAS: 'bjtobiaas',
    VDC_DTTOBIAAS: 'dttobiaas',
    VDC_JSWXTOBDDOS: 'jswxtobddos',
    VDC_LFTOBVG: 'lftobvg',
    VDC_GDSZTOBDDOS: 'gdsztobddos',
    VDC_BOEVOLC: 'boevolc',
    VDC_SHTOBOXM: 'shtoboxm',
    VDC_LFTOBAUTODRIVE: 'lftobautodrive',
    VDC_WLTOBIAAS: 'wltobiaas',
    VDC_WHTOBIAAS: 'whtobiaas',
    VDC_CDTOBIAAS: 'cdtobiaas',
    VDC_LFTOBXMDRIVE: 'lftobxmdrive',
    VDC_NMTOBDDOS: 'nmtobddos',
    VDC_SHTOBDRIVE: 'shtobdrive',
    VDC_BJTOBDDOS: 'bjtobddos',
    VDC_SINFSHA: 'sinfsha',
    VDC_SINFSHC: 'sinfshc',
    VDC_SINFVOLCVIKINGEAST: 'sinfvolcvikingeast',
    VDC_VOLCNT: 'volcnt',
    VDC_VOLCGZ: 'volcgz',
    VDC_GCPTW: 'gcptw',
    VDC_GCPHK: 'gcphk',
    VDC_SDQD: 'sdqd',
    VDC_HKCJ: 'hkcj',
    VDC_SDQDCJ: 'sdqdcj',
    VDC_JSWXMUTENS: 'jswxmutens',
    VDC_HBXYMUTENS: 'hbxymutens',
    VDC_HNCZMUTENS: 'hnczmutens',
    VDC_HBSZJMUTENS: 'hbszjmutens',
    VDC_SXXAMUTENS: 'sxxamutens',
    VDC_GDGZCMENS: 'gdgzcmens',
    VDC_GDGZCUENS: 'gdgzcuens',
    VDC_GDGZCTENS: 'gdgzctens',
    VDC_SCCDCTENS: 'sccdctens',
    VDC_SCCDCUENS: 'sccdcuens',
    VDC_SCCDCMENS: 'sccdcmens',
    VDC_HNZZCUENS: 'hnzzcuens',
    VDC_HNZZCMENS: 'hnzzcmens',
    VDC_HNZZCTENS: 'hnzzctens',
    VDC_SDJNCMENS: 'sdjncmens',
    VDC_ZJHZCMENS: 'zjhzcmens',
    VDC_SDJNCUENS: 'sdjncuens',
    VDC_ZJTZCUENS: 'zjtzcuens',
    VDC_ZJHZCTENS: 'zjhzctens',
    VDC_SHCUENS: 'shcuens',
    VDC_SHCMENS: 'shcmens',
    VDC_SHCTENS: 'shctens',
    VDC_GXNNCUENS: 'gxnncuens',
    VDC_GXNNCMENS: 'gxnncmens',
    VDC_GXNNCTENS: 'gxnnctens',
    VDC_SXXACMENS: 'sxxacmens',
    VDC_AHWHCTCDN: 'ahwhctcdn',
    VDC_SDQDCTENS: 'sdqdctens',
    VDC_GSLZCMENS: 'gslzcmens',
    VDC_GSLZCTENS: 'gslzctens',
    VDC_NMGHHHTCTENS: 'nmghhhtctens',
    VDC_GSLZCUENS: 'gslzcuens',
    VDC_NMGHHHTCMENS: 'nmghhhtcmens',
    VDC_NMGHHHTCUENS: 'nmghhhtcuens',
    VDC_TJCUENS: 'tjcuens',
    VDC_TJCMENS: 'tjcmens',
    VDC_TJCTENS: 'tjctens',
    VDC_HLJHEBCTENS: 'hljhebctens',
    VDC_HLJHEBCMENS: 'hljhebcmens',
    VDC_HLJHEBCUENS: 'hljhebcuens',
    VDC_JLCHCMENS: 'jlchcmens',
    VDC_JLJLCUENS: 'jljlcuens',
    VDC_JLTHCTENS: 'jlthctens',
    VDC_XJWLMQCMENS: 'xjwlmqcmens',
    VDC_XJWLMQCTENS: 'xjwlmqctens',
    VDC_XJWLMQCUENS: 'xjwlmqcuens',
    VDC_CQCMENS: 'cqcmens',
    VDC_CQCUENS: 'cqcuens',
    VDC_CQCTENS: 'cqctens',
    VDC_BJCTENS: 'bjctens',
    VDC_BJCMENS: 'bjcmens',
    VDC_BJCUENS: 'bjcuens',
    VDC_LNDLCMENS: 'lndlcmens',
    VDC_LNDLCUENS: 'lndlcuens',
    VDC_SXTYCTENS: 'sxtyctens',
    VDC_FJXMMUTENS: 'fjxmmutens',
    VDC_SXTYCUENS: 'sxtycuens',
    VDC_YNKMCTENS: 'ynkmctens',
    VDC_JXNCCTENS: 'jxncctens',
    VDC_JXNCCMENS: 'jxnccmens',
    VDC_GZGYCMENS: 'gzgycmens',
    VDC_GZGYCUENS: 'gzgycuens',
    VDC_GZGYCTENS: 'gzgyctens',
    VDC_YNKMCUENS: 'ynkmcuens',
    VDC_YNKMCMENS: 'ynkmcmens',
    VDC_LNDLCTENS: 'lndlctens',
    VDC_JXNCCUENS: 'jxnccuens',
    VDC_HNHKCUENS: 'hnhkcuens',
    VDC_XZLSCUENS: 'xzlscuens',
    VDC_GDFSCTENS: 'gdfsctens',
    VDC_HLJJMSCUENS: 'hljjmscuens',
    VDC_HNCSCUENS: 'hncscuens',
    VDC_GDDGCTENS: 'gddgctens',
    VDC_HBWHCTENS: 'hbwhctens',
    VDC_HBWHCUENS: 'hbwhcuens',
    VDC_HNHKCTENS: 'hnhkctens',
    VDC_JSSZCTENS: 'jsszctens',
    VDC_FJFZCTENS: 'fjfzctens',
    VDC_SXTYCMENS: 'sxtycmens',
    VDC_XZLSCMENS: 'xzlscmens',
    VDC_QHXNCMENS: 'qhxncmens',
    VDC_NXYCCMENS: 'nxyccmens',
    VDC_NXZWCTENS: 'nxzwctens',
    VDC_AHHFCMENS: 'ahhfcmens',
    VDC_NXYCCUENS: 'nxyccuens',
    VDC_HNHKCMENS: 'hnhkcmens',
    VDC_AHBBCTENS: 'ahbbctens',
    VDC_HNCSCTENS: 'hncsctens',
    VDC_JSSZCMENS: 'jsszcmens',
    VDC_JSNJCMENS: 'jsnjcmens',
    VDC_JSNJCUENS: 'jsnjcuens',
    VDC_SDQDCUENS: 'sdqdcuens',
    VDC_AAA: 'aaa',
    VDC_ALIMUTCNZJHZ01: 'alimutcnzjhz01',
    VDC_ALIMUTCNHBWH01: 'alimutcnhbwh01',
    VDC_ALICMCNJXNC04: 'alicmcnjxnc04',
    VDC_ALICMCNSDJN02: 'alicmcnsdjn02',
    VDC_ALICMCNHBWH02: 'alicmcnhbwh02',
    VDC_ALICMCNHAZZ06: 'alicmcnhazz06',
    VDC_ALICMCNHBWH01: 'alicmcnhbwh01',
    VDC_ALICTCNGXNN02: 'alictcngxnn02',
    VDC_ALICMCNHAZZ05: 'alicmcnhazz05',
    VDC_ALICTCNFJQZ01: 'alictcnfjqz01',
    VDC_ALICTCNHBWH04: 'alictcnhbwh04',
    VDC_ALICTCNSDQD02: 'alictcnsdqd02',
    VDC_ALICMCNJXNC03: 'alicmcnjxnc03',
    VDC_ALICMCNYNKM04: 'alicmcnynkm04',
    VDC_ALICMCNGXNN03: 'alicmcngxnn03',
    VDC_ALICMCNGXNN02: 'alicmcngxnn02',
    VDC_ALICTCNHBWH03: 'alictcnhbwh03',
    VDC_ALICMCNHNCS03: 'alicmcnhncs03',
    VDC_ALICMCNHEBD02: 'alicmcnhebd02',
    VDC_ALICMCNLNDL03: 'alicmcnlndl03',
    VDC_ALICTCNGDSZ02: 'alictcngdsz02',
    VDC_ALICMCNZJNB01: 'alicmcnzjnb01',
    VDC_ALICMCNJSNJ02: 'alicmcnjsnj02',
    VDC_ALICMCNSCCD06: 'alicmcnsccd06',
    VDC_ALICMCNFJXM02: 'alicmcnfjxm02',
    VDC_ALICMCNFJXM01: 'alicmcnfjxm01',
    VDC_ALICMCNHAZZ03: 'alicmcnhazz03',
    VDC_ALICMCNHAZZ04: 'alicmcnhazz04',
    VDC_ALICMCNXJWLMQ02: 'alicmcnxjwlmq02',
    VDC_ALICMCNLNDL02: 'alicmcnlndl02',
    VDC_ALICMCNYNKM03: 'alicmcnynkm03',
    VDC_ALICMCNSCCD05: 'alicmcnsccd05',
    VDC_ALICMCNYNKM02: 'alicmcnynkm02',
    VDC_ALICMCNSCCD04: 'alicmcnsccd04',
    VDC_ALICMCNHBXY02: 'alicmcnhbxy02',
    VDC_ALICMCNHIHK02: 'alicmcnhihk02',
    VDC_ALICMCNGZGY03: 'alicmcngzgy03',
    VDC_ALICMCNGZGY02: 'alicmcngzgy02',
    VDC_ALICMCNGDGZ02: 'alicmcngdgz02',
    VDC_ALICMCNHEBD01: 'alicmcnhebd01',
    VDC_ALICMCNSCCD03: 'alicmcnsccd03',
    VDC_ALICTCNGZGY02: 'alictcngzgy02',
    VDC_ALICMCNHAZZ02: 'alicmcnhazz02',
    VDC_ALICMCNSCCD02: 'alicmcnsccd02',
    VDC_HBSJZMUTENS: 'hbsjzmutens',
    VDC_ALICUCNHESJZ02: 'alicucnhesjz02',
    VDC_ALICMCNZJHZ03: 'alicmcnzjhz03',
    VDC_ALICUCNGDDG01: 'alicucngddg01',
    VDC_ALICMCNHNCS02: 'alicmcnhncs02',
    VDC_ALICTCNJSWX01: 'alictcnjswx01',
    VDC_ALICTCNZJHZ01: 'alictcnzjhz01',
    VDC_ALICMCNZJHZ02: 'alicmcnzjhz02',
    VDC_ALICTCNQHXN01: 'alictcnqhxn01',
    VDC_ALICUCNAHWH01: 'alicucnahwh01',
    VDC_ALICTCNSNXA02: 'alictcnsnxa02',
    VDC_ALICTCNHBWH02: 'alictcnhbwh02',
    VDC_ALICUCNHLSH01: 'alicucnhlsh01',
    VDC_ALICTCNSHSH03: 'alictcnshsh03',
    VDC_ALICUCNSDQD01: 'alicucnsdqd01',
    VDC_ALICTCNSCCD04: 'alictcnsccd04',
    VDC_ALICMCNHAZZ07: 'alicmcnhazz07',
    VDC_ALICUCNSXTY02: 'alicucnsxty02',
    VDC_ALICTCNGDFS03: 'alictcngdfs03',
    VDC_ALICMCNZJNB02: 'alicmcnzjnb02',
    VDC_ALICMCNSCCD07: 'alicmcnsccd07',
    VDC_ALICMCNYNKM05: 'alicmcnynkm05',
    VDC_ALICTCNAHHF01: 'alictcnahhf01',
    VDC_ALICUCNGZGY02: 'alicucngzgy02',
    VDC_ALICTCNSDJN02: 'alictcnsdjn02',
    VDC_ALICMCNGXLZ01: 'alicmcngxlz01',
    VDC_ALICMCNGSLZ03: 'alicmcngslz03',
    VDC_ALICTCNBJBJ02: 'alictcnbjbj02',
    VDC_ALICTCNSCYB01: 'alictcnscyb01',
    VDC_ALIMUTCNSCCD01: 'alimutcnsccd01',
    VDC_ALICTCNSCYB02: 'alictcnscyb02',
    VDC_ALIMUTCNJXNC01: 'alimutcnjxnc01',
    VDC_ALIMUTCNGDFS01: 'alimutcngdfs01',
    VDC_ALIMUTCNGDDG01: 'alimutcngddg01',
    VDC_ALIMUTCNHAZZ01: 'alimutcnhazz01',
    VDC_ALIMUTCNLNDL01: 'alimutcnlndl01',
    VDC_ALIMUTCNTJTJ01: 'alimutcntjtj01',
    VDC_ALIMUTCNSXTY01: 'alimutcnsxty01',
    VDC_ALICTCNSXTY02: 'alictcnsxty02',
    VDC_ALIMUTCNJXNC02: 'alimutcnjxnc02',
    VDC_ALICTCNQHHD01: 'alictcnqhhd01',
    VDC_ALIMUTCNGZGY02: 'alimutcngzgy02',
    VDC_ALICMCNYNKM06: 'alicmcnynkm06',
    VDC_ALIMUTCNHNLD01: 'alimutcnhnld01',
    VDC_ALIMUTCNHESJZ02: 'alimutcnhesjz02',
    VDC_ALICTCNHELF01: 'alictcnhelf01',
    VDC_ALICUCNSDQD03: 'alicucnsdqd03',
    VDC_ALIMUTCNJXNC03: 'alimutcnjxnc03',
    VDC_ALIMUTCNHALY01: 'alimutcnhaly01',
    VDC_ALICUCNGXLZ01: 'alicucngxlz01',
    VDC_ALICUCNGSLZ03: 'alicucngslz03',
    VDC_ALIMUTCNHESJZ03: 'alimutcnhesjz03',
    VDC_ALIMUTCNGDDG02: 'alimutcngddg02',
    VDC_ALIMUTCNHBXY02: 'alimutcnhbxy02',
    VDC_ALICMCNLNDL04: 'alicmcnlndl04',
    VDC_ALICTCNNXZW03: 'alictcnnxzw03',
    VDC_ALICTCNSDQD03: 'alictcnsdqd03',
    VDC_ALIMUTCNFJQZ01: 'alimutcnfjqz01',
    VDC_ALICMCNLNDL05: 'alicmcnlndl05',
    VDC_ALICTCNSXTY03: 'alictcnsxty03',
    VDC_ALICTCNNMHHHT02: 'alictcnnmhhht02',
    VDC_ALIMUTCNSCDY01: 'alimutcnscdy01',
    VDC_ALIMUTCNHESJZ04: 'alimutcnhesjz04',
    VDC_ALIMUTCNGDGZ01: 'alimutcngdgz01',
    VDC_ALIMUTCNZJJX01: 'alimutcnzjjx01',
    VDC_ALICUCNZJJH01: 'alicucnzjjh01',
    VDC_ALICMCNLNSY01: 'alicmcnlnsy01',
    VDC_ALIMUTCNAHHF01: 'alimutcnahhf01',
    VDC_ALIMUTCNHBWH02: 'alimutcnhbwh02',
    VDC_ALICTCNSDQD04: 'alictcnsdqd04',
    VDC_ALICUCNSNXA02: 'alicucnsnxa02',
    VDC_ALIMUTCNLNSY01: 'alimutcnlnsy01',
    VDC_ALICUCNJLCC02: 'alicucnjlcc02',
    VDC_ALICMCNJSNJ03: 'alicmcnjsnj03',
    VDC_ALIMUTCNYNKM01: 'alimutcnynkm01',
    VDC_ALICUCNYNKM03: 'alicucnynkm03',
    VDC_ALICTCNBJBJ03: 'alictcnbjbj03',
    VDC_ALICUCNZJJH02: 'alicucnzjjh02',
    VDC_ALICUCNJSCZ01: 'alicucnjscz01',
    VDC_ALICUCNLNSY01: 'alicucnlnsy01',
    VDC_ALIMUTCNZJJH01: 'alimutcnzjjh01',
    VDC_ALICMCNLNDL06: 'alicmcnlndl06',
    VDC_ALICTCNSDYT01: 'alictcnsdyt01',
    VDC_ALICMCNNXZW01: 'alicmcnnxzw01',
    VDC_ALICTCNGDGZ03: 'alictcngdgz03',
    VDC_ALICMCNGZGY04: 'alicmcngzgy04',
    VDC_ALICTCNCQCQ02: 'alictcncqcq02',
    VDC_ALICMCNHLHEB02: 'alicmcnhlheb02',
    VDC_ALICMCNTJTJ02: 'alicmcntjtj02',
    VDC_ALICMCNGZGY05: 'alicmcngzgy05',
    VDC_ALICUCNCQCQ02: 'alicucncqcq02',
    VDC_ALIMUTCNAHHF02: 'alimutcnahhf02',
    VDC_ALIMUTCNZJWZ01: 'alimutcnzjwz01',
    VDC_ALICUCNSNXY01: 'alicucnsnxy01',
    VDC_ALICMCNHAZZ08: 'alicmcnhazz08',
    VDC_ALICMCNSXTY02: 'alicmcnsxty02',
    VDC_ALICMCNSNBJ01: 'alicmcnsnbj01',
    VDC_ALICTCNGXNN03: 'alictcngxnn03',
    VDC_CDN: 'cdn',
    VDC_GALINC2: 'galinc2',
    VDC_BOE: 'boe',
    VDC_COF: 'cof',
    VDC_DEVBOX: 'devbox',
    VDC_BOETEST: 'boetest',
    VDC_BOE2: 'boe2',
    VDC_SINFBOE: 'sinfboe',
    VDC_DEVBOXI18N: 'devboxi18n',
    VDC_BOEI18N: 'boei18n',
    VDC_BOETTP: 'boettp',
    VDC_BOEVPC2: 'boevpc2',
    VDC_BOESG: 'boesg',
    VDC_DEVBOXSG: 'devboxsg',
    VDC_BOEINF: 'boeinf',
    VDC_AUTODEPLOYTEST: 'autodeploytest',
    VDC_AUTODEPLOYCICD: 'autodeploycicd',
    VDC_MALIVA: 'maliva',
    VDC_USEAST1A: 'useast1a',
    VDC_USEAST3: 'useast3',
    VDC_USEAST4: 'useast4',
    VDC_GCPBR: 'gcpbr',
    VDC_AWSVAGM: 'awsvagm',
    VDC_AWSVAC: 'awsvac',
    VDC_GCPUSIAD: 'gcpusiad',
    VDC_ALIVA: 'aliva',
    VDC_USEASTDT: 'useastdt',
    VDC_USEASTAZURE: 'useastazure',
    VDC_USTSENTRY: 'ustsentry',
    VDC_USEAST6: 'useast6',
    VDC_USEAST7: 'useast7',
    VDC_USEAST10A: 'useast10a',
    VDC_USEASTIC: 'useastic',
    VDC_USEAST2AZURE: 'useast2azure',
    VDC_USEAST1B: 'useast1b',
    VDC_ALISG: 'alisg',
    VDC_SG: 'sg',
    VDC_SG1: 'sg1',
    VDC_SGDT: 'sgdt',
    VDC_SGAZURE: 'sgazure',
    VDC_SGISOLATION: 'sgisolation',
    VDC_SG2: 'sg2',
    VDC_MY: 'my',
    VDC_MY2: 'my2',
    VDC_MY3: 'my3',
    VDC_SGSAAS1LARKIDC1: 'sgsaas1larkidc1',
    VDC_SGSAAS1LARKIDC2: 'sgsaas1larkidc2',
    VDC_SGSAAS1LARKIDC3: 'sgsaas1larkidc3',
    VDC_SGCOMPLIANCE: 'sgcompliance',
    VDC_MYA: 'mya',
    VDC_MYB: 'myb',
    VDC_SG3A: 'sg3a',
    VDC_ID3B: 'id3b',
    VDC_HK1A: 'hk1a',
    VDC_MYC: 'myc',
    VDC_BDSGDT: 'bdsgdt',
    VDC_MY5A: 'my5a',
    VDC_ID4B: 'id4b',
    VDC_ID4A: 'id4a',
    VDC_MYCISA: 'mycisa',
    VDC_MYCISB: 'mycisb',
    VDC_SGCISA: 'sgcisa',
    VDC_JPSAAS: 'jpsaas',
    VDC_JPSAAS2A: 'jpsaas2a',
    VDC_AWSFR: 'awsfr',
    VDC_GCPPL: 'gcppl',
    VDC_FRACOMPLIANCE: 'fracompliance',
    VDC_BE1A: 'be1a',
    VDC_FR1A: 'fr1a',
    VDC_ID1A: 'id1a',
    VDC_ID2A: 'id2a',
    VDC_USEAST11A: 'useast11a',
    VDC_USWEST1A: 'uswest1a',
    VDC_CA: 'ca',
    VDC_AWSUSWEST2: 'awsuswest2',
    VDC_USWEST2: 'uswest2',
    VDC_USWEST3A: 'uswest3a',
    VDC_USWEST3B: 'uswest3b',
    VDC_USSW1B: 'ussw1b',
    VDC_USSW1A: 'ussw1a',
    VDC_GCPUSCBF: 'gcpuscbf',
    VDC_GCPUNUSE: 'gcpunuse',
    VDC_USORDX: 'usordx',
    VDC_USCENTRAL1A: 'uscentral1a',
    VDC_VA: 'va',
    VDC_USEAST2A: 'useast2a',
    VDC_USEAST2B: 'useast2b',
    VDC_USEAST9A: 'useast9a',
    VDC_USEAST12A: 'useast12a',
    VDC_USEAST13A: 'useast13a',
    VDC_USEAST14A: 'useast14a',
    VDC_GCPCA: 'gcpca',
    VDC_SGCOMM1: 'sgcomm1',
    VDC_SGEA2: 'sgea2',
    VDC_SGEA1: 'sgea1',
    VDC_SGEE2: 'sgee2',
    VDC_SGEE1: 'sgee1',
    VDC_SGEE3: 'sgee3',
    VDC_GALISG: 'galisg',
    VDC_GCPIN: 'gcpin',
    VDC_AWSIN: 'awsin',
    VDC_IND: 'ind',
    VDC_INDDP: 'inddp',
    VDC_GCPINDEL: 'gcpindel',
    VDC_JOHORTOBIAAS: 'johortobiaas',
    VDC_HKTOBIAAS: 'hktobiaas',
    VDC_SGTOBIAAS: 'sgtobiaas',
    VDC_JKTTOBIAAS: 'jkttobiaas',
    VDC_GCPJPOSA: 'gcpjposa',
    VDC_GCPJPTKY: 'gcpjptky',
    VDC_AWSJP: 'awsjp',
    VDC_GCPSG: 'gcpsg',
    VDC_AWSSG: 'awssg',
    VDC_GCPID: 'gcpid',
    VDC_AWSIDHLP: 'awsidhlp',
    VDC_AWSKR: 'awskr',
    VDC_ASID1A: 'asid1a',
    VDC_ASID2A: 'asid2a',
    VDC_AWSDEGM: 'awsdegm',
    VDC_AWSJPGM: 'awsjpgm',
    VDC_AWSSGGM: 'awssggm',
    VDC_PROGAME: 'progame',
    VDC_GCPKR: 'gcpkr',
    VDC_BGE: 'bge',
    VDC_GCPSG2: 'gcpsg2',
    VDC_GCPAU: 'gcpau',
    VDC_SYD1A: 'syd1a',
    VDC_YCRU: 'ycru',
    VDC_DUBCOMPLIANCE: 'dubcompliance',
    VDC_FR: 'fr',
    VDC_KSRU: 'ksru',
    VDC_GCPFI: 'gcpfi',
    VDC_GCPBE: 'gcpbe',
    VDC_GCPGB: 'gcpgb',
    VDC_GCPDE: 'gcpde',
    VDC_GCPNL: 'gcpnl',
    VDC_GCPCH: 'gcpch',
    VDC_AWSUKLD: 'awsukld',
    VDC_AWSBH: 'awsbh',
    VDC_RTCSA: 'rtcsa',
    VDC_AWSBR: 'awsbr',
    VDC_AWSSFCPT: 'awssfcpt',
    VDC_USEAST5: 'useast5',
    VDC_USEAST8: 'useast8',
    VDC_IE: 'ie',
    VDC_DEDT: 'dedt',
    VDC_IEDT: 'iedt',
    VDC_EUIE1A: 'euie1a',
    VDC_IEAZURE: 'ieazure',
    VDC_NO: 'no',
    VDC_NO1A: 'no1a',
    VDC_NOAZURE: 'noazure',
    VDC_DE: 'de',
    VDC_IE2: 'ie2',
    VDC_JPSOUTH3A: 'jpsouth3a',
    VDC_DY: 'dy',
    VDC_YX: 'yx',
    VDC_ZC: 'zc',
    VDC_VATOBIAAS: 'vatobiaas',
    VDC_HLHKPAY: 'hlhkpay',
    VDC_LFHKPAY: 'lfhkpay',
    VDC_IE3A: 'ie3a',
    VDC_BDDEDT: 'bddedt',
    VDC_BE2A: 'be2a',
    VDC_DE2A: 'de2a',
    VDC_MY4A: 'my4a',
    VDC_WLBY: 'wlby',
    VDC_GL2: 'gl2',
    VDC_XH: 'xh',
    VDC_ZB: 'zb',
    VDC_HLTOBIAASBOE: 'hltobiaasboe',
    VDC_BGE2: 'bge2',
    VDC_BGEEU: 'bgeeu',
    VDC_SYD2A: 'syd2a',
    VDC_USEAST15A: 'useast15a',
    VDC_TH1A: 'th1a',
    VDC_TH1B: 'th1b',
    VDC_USSOUTHWEST1A: 'ussw1a',
    VDC_USSOUTHWEST1B: 'ussw1b',
};
exports.VREGION = {
    VREGION_UNKNOWN: '-',
    VREGION_CHINANORTH: 'China-North',
    VREGION_CHINAEAST: 'China-East',
    VREGION_CHINASINFNORTH: 'ChinaSinf-North',
    VREGION_CHINAAGGREGATION: 'China-Aggregation',
    VREGION_CHINAPAY: 'China-Pay',
    VREGION_CHINAPAY2: 'China-Pay2',
    VREGION_CHINAENTERPRISE: 'China-Enterprise',
    VREGION_CHINASINFEAST: 'ChinaSinf-East',
    VREGION_CHINAVOLCEAST: 'ChinaVolc-East',
    VREGION_CHINAVOLCSOUTH: 'ChinaVolc-South',
    VREGION_CHINATW: 'China-TW',
    VREGION_CHINAHK: 'China-HK',
    VREGION_CHINAFINTECH: 'China-Fintech',
    VREGION_CHINAEDGE: 'China-Edge',
    VREGION_CHINACDN: 'China-CDN',
    VREGION_ALIYUNNC2: 'Aliyun_NC2',
    VREGION_CHINABOE: 'China-BOE',
    VREGION_CHINABOE2: 'China-BOE2',
    VREGION_CHINASINFBOE: 'ChinaSinf-BOE',
    VREGION_USBOE: 'US-BOE',
    VREGION_SINGAPOREBOE: 'Singapore-BOE',
    VREGION_CHINAINFBOE: 'China-InfBOE',
    VREGION_USEAST: 'US-East',
    VREGION_SINGAPORECENTRAL: 'Singapore-Central',
    VREGION_SINGAPORESAAS: 'Singapore-SaaS',
    VREGION_SINGAPORECOMPLIANCE: 'Singapore-Compliance',
    VREGION_ASIASOUTHEASTBD: 'Asia-SouthEastBD',
    VREGION_ASIACIS: 'Asia-CIS',
    VREGION_ASIASAAS: 'Asia-SaaS',
    VREGION_EUROPECENTRAL: 'Europe-Central',
    VREGION_IDCOMPLIANCE: 'ID-Compliance',
    VREGION_IDCOMPLIANCE2: 'ID-Compliance2',
    VREGION_USCOMPLIANCE: 'US-Compliance',
    VREGION_USWEST: 'US-West',
    VREGION_USSOUTHWEST: 'US-SouthWest',
    VREGION_USCENTRAL: 'US-Central',
    VREGION_USEE: 'US-EE',
    VREGION_USEASTRED: 'US-EastRed',
    VREGION_USEASTBD: 'US-EastBD',
    VREGION_USNORTHEAST: 'US-NorthEast',
    VREGION_SINGAPORECOMMON: 'Singapore-Common',
    VREGION_SINGAPOREEE: 'Singapore-EE',
    VREGION_SINGAPOREENTERPRISE: 'Singapore-Enterprise',
    VREGION_SINGAPOREGAME: 'Singapore-Game',
    VREGION_ASIASOUTH: 'Asia-South',
    VREGION_ASIACOMPLIANCE: 'Asia-Compliance',
    VREGION_ASIACOMPLIANCE2: 'Asia-Compliance2',
    VREGION_ASIAEAST: 'Asia-East',
    VREGION_ASIAENTERPRISE: 'Asia-Enterprise',
    VREGION_ASIANORTHEAST: 'Asia-NorthEast',
    VREGION_ASIASOUTHEAST: 'Asia-SouthEast',
    VREGION_I18NGAME: 'I18N-Game',
    VREGION_I18NBGE: 'I18N-BGE',
    VREGION_AUSTRALIASOUTHEAST: 'Australia-SouthEast',
    VREGION_EASTERNEUROTT: 'EasternEuro-TT',
    VREGION_EUROPECOMPLIANCE: 'Europe-Compliance',
    VREGION_EUROPEEAST: 'Europe-East',
    VREGION_EUROPENORTH: 'Europe-North',
    VREGION_EUROPEWEST: 'Europe-West',
    VREGION_MIDDLEEASTSOUTH: 'MiddleEast-South',
    VREGION_SOUTHAMERICAEAST: 'SouthAmerica-East',
    VREGION_AFRICASOUTH: 'Africa-South',
    VREGION_USTTP: 'US-TTP',
    VREGION_USTTP2: 'US-TTP2',
    VREGION_EUTTP: 'EU-TTP',
    VREGION_EUTTP2: 'EU-TTP2',
    VREGION_EUCOMPLIANCE: 'EU-Compliance',
    VREGION_EUCOMPLIANCE2: 'EU-Compliance2',
    VREGION_ASIANORTHEASTBD: 'Asia-NorthEastBD',
    VREGION_CHINANORTH3: 'China-North3',
    VREGION_USENTERPRISE: 'US-Enterprise',
    VREGION_CHINAHKPAY: 'China-HKPay',
    VREGION_EUROPEWESTBD: 'Europe-WestBD',
    VREGION_MYCOMPLIANCE: 'MY-Compliance',
    VREGION_CHINANORTH5: 'China-North5',
    VREGION_USEAST2: 'US-East2',
    VREGION_USWESTBD: 'US-WestBD',
    VREGION_EUROPEEASTBD: 'Europe-EastBD',
    VREGION_CHINANORTH6: 'China-North6',
    VREGION_CHINAENTERPRISEBOE: 'China-EnterpriseBOE',
    VREGION_EUBGE: 'EU-BGE',
    VREGION_AUSTRALIASOUTHEASTBD: 'Australia-SouthEastBD',
    VREGION_CHINANORTH2: 'China-North2',
    VREGION_USTTP3: 'US-TTP3',
    VREGION_THCENTRAL: 'TH-Central',
    VREGION_SINFCHINAEAST: 'ChinaSinf-East',
    VREGION_SINFCHINANORTH: 'ChinaSinf-North',
};
exports.VDC_TO_VREGION_MAP = {
    [exports.VDC.VDC_UNKNOWN]: exports.VREGION.VREGION_UNKNOWN,
    [exports.VDC.VDC_LF]: exports.VREGION.VREGION_CHINANORTH,
    [exports.VDC.VDC_HL]: exports.VREGION.VREGION_CHINANORTH,
    [exports.VDC.VDC_LQ]: exports.VREGION.VREGION_CHINANORTH,
    [exports.VDC.VDC_WJ]: exports.VREGION.VREGION_CHINANORTH,
    [exports.VDC.VDC_AWSNC1]: exports.VREGION.VREGION_CHINANORTH,
    [exports.VDC.VDC_YG]: exports.VREGION.VREGION_CHINANORTH,
    [exports.VDC.VDC_ZK]: exports.VREGION.VREGION_CHINANORTH,
    [exports.VDC.VDC_GL]: exports.VREGION.VREGION_CHINANORTH,
    [exports.VDC.VDC_TJDT]: exports.VREGION.VREGION_CHINANORTH,
    [exports.VDC.VDC_PD]: exports.VREGION.VREGION_CHINAEAST,
    [exports.VDC.VDC_HJ]: exports.VREGION.VREGION_CHINAEAST,
    [exports.VDC.VDC_PDDT]: exports.VREGION.VREGION_CHINAEAST,
    [exports.VDC.VDC_YZ]: exports.VREGION.VREGION_CHINAEAST,
    [exports.VDC.VDC_ZJG]: exports.VREGION.VREGION_CHINAEAST,
    [exports.VDC.VDC_JJ]: exports.VREGION.VREGION_CHINAEAST,
    [exports.VDC.VDC_YJ]: exports.VREGION.VREGION_CHINAEAST,
    [exports.VDC.VDC_SINFONLINE]: exports.VREGION.VREGION_CHINASINFNORTH,
    [exports.VDC.VDC_SINFONLINEA]: exports.VREGION.VREGION_CHINASINFNORTH,
    [exports.VDC.VDC_SINFONLINEC]: exports.VREGION.VREGION_CHINASINFNORTH,
    [exports.VDC.VDC_SINFVOLCVIKING]: exports.VREGION.VREGION_CHINASINFNORTH,
    [exports.VDC.VDC_AGCQ]: exports.VREGION.VREGION_CHINAAGGREGATION,
    [exports.VDC.VDC_AGGDSZ]: exports.VREGION.VREGION_CHINAAGGREGATION,
    [exports.VDC.VDC_AGHBWH]: exports.VREGION.VREGION_CHINAAGGREGATION,
    [exports.VDC.VDC_AGJSNJ]: exports.VREGION.VREGION_CHINAAGGREGATION,
    [exports.VDC.VDC_AGLNSY]: exports.VREGION.VREGION_CHINAAGGREGATION,
    [exports.VDC.VDC_AGSDQD]: exports.VREGION.VREGION_CHINAAGGREGATION,
    [exports.VDC.VDC_AGSXLQ]: exports.VREGION.VREGION_CHINAAGGREGATION,
    [exports.VDC.VDC_AGSXXA]: exports.VREGION.VREGION_CHINAAGGREGATION,
    [exports.VDC.VDC_AGSY]: exports.VREGION.VREGION_CHINAAGGREGATION,
    [exports.VDC.VDC_ALIEC1]: exports.VREGION.VREGION_CHINAAGGREGATION,
    [exports.VDC.VDC_ALIEC2]: exports.VREGION.VREGION_CHINAAGGREGATION,
    [exports.VDC.VDC_ALINC2]: exports.VREGION.VREGION_CHINAAGGREGATION,
    [exports.VDC.VDC_ALIGDSZ]: exports.VREGION.VREGION_CHINAAGGREGATION,
    [exports.VDC.VDC_ALISH]: exports.VREGION.VREGION_CHINAAGGREGATION,
    [exports.VDC.VDC_ALISC1]: exports.VREGION.VREGION_CHINAAGGREGATION,
    [exports.VDC.VDC_BJLGY]: exports.VREGION.VREGION_CHINAAGGREGATION,
    [exports.VDC.VDC_ALINC2HL]: exports.VREGION.VREGION_CHINAAGGREGATION,
    [exports.VDC.VDC_ALINC2LF]: exports.VREGION.VREGION_CHINAAGGREGATION,
    [exports.VDC.VDC_ALINC2LQ]: exports.VREGION.VREGION_CHINAAGGREGATION,
    [exports.VDC.VDC_LFZG]: exports.VREGION.VREGION_CHINAPAY,
    [exports.VDC.VDC_HLZG]: exports.VREGION.VREGION_CHINAPAY,
    [exports.VDC.VDC_HJZG]: exports.VREGION.VREGION_CHINAPAY2,
    [exports.VDC.VDC_QDTOB]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_QDTOBIAAS]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_SYKA2LARK]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_SYKALARK]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_SYKA3LARK]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_LFTOBIAAS]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_HLTOBIAAS]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_GZTOBIAAS]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_NTTOBIAAS]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_ZJJHTOBDDOS]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_HBSJZTOBDDOS]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_SHTOBIAAS]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_JSYZTOBDDOS]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_BOEXFL]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_GDDGTOBDDOS]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_XFLONLINE]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_BOETOBIAAS]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_LFTOBFINANCE]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_SHTOBFINANCE]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_ZJTZTOBDDOS]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_BJTOBIAAS]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_DTTOBIAAS]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_JSWXTOBDDOS]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_LFTOBVG]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_GDSZTOBDDOS]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_BOEVOLC]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_SHTOBOXM]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_LFTOBAUTODRIVE]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_WLTOBIAAS]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_WHTOBIAAS]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_CDTOBIAAS]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_LFTOBXMDRIVE]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_NMTOBDDOS]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_SHTOBDRIVE]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_BJTOBDDOS]: exports.VREGION.VREGION_CHINAENTERPRISE,
    [exports.VDC.VDC_SINFSHA]: exports.VREGION.VREGION_CHINASINFEAST,
    [exports.VDC.VDC_SINFSHC]: exports.VREGION.VREGION_CHINASINFEAST,
    [exports.VDC.VDC_SINFVOLCVIKINGEAST]: exports.VREGION.VREGION_CHINASINFEAST,
    [exports.VDC.VDC_VOLCNT]: exports.VREGION.VREGION_CHINAVOLCEAST,
    [exports.VDC.VDC_VOLCGZ]: exports.VREGION.VREGION_CHINAVOLCSOUTH,
    [exports.VDC.VDC_GCPTW]: exports.VREGION.VREGION_ASIAEAST,
    [exports.VDC.VDC_GCPHK]: exports.VREGION.VREGION_ASIAEAST,
    [exports.VDC.VDC_SDQD]: exports.VREGION.VREGION_CHINAFINTECH,
    [exports.VDC.VDC_HKCJ]: exports.VREGION.VREGION_CHINAFINTECH,
    [exports.VDC.VDC_SDQDCJ]: exports.VREGION.VREGION_CHINAFINTECH,
    [exports.VDC.VDC_JSWXMUTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_HBXYMUTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_HNCZMUTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_HBSZJMUTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_SXXAMUTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_GDGZCMENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_GDGZCUENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_GDGZCTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_SCCDCTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_SCCDCUENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_SCCDCMENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_HNZZCUENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_HNZZCMENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_HNZZCTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_SDJNCMENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ZJHZCMENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_SDJNCUENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ZJTZCUENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ZJHZCTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_SHCUENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_SHCMENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_SHCTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_GXNNCUENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_GXNNCMENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_GXNNCTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_SXXACMENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_AHWHCTCDN]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_SDQDCTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_GSLZCMENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_GSLZCTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_NMGHHHTCTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_GSLZCUENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_NMGHHHTCMENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_NMGHHHTCUENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_TJCUENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_TJCMENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_TJCTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_HLJHEBCTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_HLJHEBCMENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_HLJHEBCUENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_JLCHCMENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_JLJLCUENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_JLTHCTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_XJWLMQCMENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_XJWLMQCTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_XJWLMQCUENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_CQCMENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_CQCUENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_CQCTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_BJCTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_BJCMENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_BJCUENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_LNDLCMENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_LNDLCUENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_SXTYCTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_FJXMMUTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_SXTYCUENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_YNKMCTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_JXNCCTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_JXNCCMENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_GZGYCMENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_GZGYCUENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_GZGYCTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_YNKMCUENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_YNKMCMENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_LNDLCTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_JXNCCUENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_HNHKCUENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_XZLSCUENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_GDFSCTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_HLJJMSCUENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_HNCSCUENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_GDDGCTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_HBWHCTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_HBWHCUENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_HNHKCTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_JSSZCTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_FJFZCTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_SXTYCMENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_XZLSCMENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_QHXNCMENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_NXYCCMENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_NXZWCTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_AHHFCMENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_NXYCCUENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_HNHKCMENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_AHBBCTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_HNCSCTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_JSSZCMENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_JSNJCMENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_JSNJCUENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_SDQDCUENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_AAA]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALIMUTCNZJHZ01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALIMUTCNHBWH01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNJXNC04]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNSDJN02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNHBWH02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNHAZZ06]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNHBWH01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNGXNN02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNHAZZ05]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNFJQZ01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNHBWH04]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNSDQD02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNJXNC03]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNYNKM04]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNGXNN03]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNGXNN02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNHBWH03]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNHNCS03]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNHEBD02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNLNDL03]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNGDSZ02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNZJNB01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNJSNJ02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNSCCD06]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNFJXM02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNFJXM01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNHAZZ03]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNHAZZ04]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNXJWLMQ02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNLNDL02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNYNKM03]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNSCCD05]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNYNKM02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNSCCD04]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNHBXY02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNHIHK02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNGZGY03]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNGZGY02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNGDGZ02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNHEBD01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNSCCD03]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNGZGY02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNHAZZ02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNSCCD02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_HBSJZMUTENS]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICUCNHESJZ02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNZJHZ03]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICUCNGDDG01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNHNCS02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNJSWX01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNZJHZ01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNZJHZ02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNQHXN01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICUCNAHWH01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNSNXA02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNHBWH02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICUCNHLSH01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNSHSH03]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICUCNSDQD01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNSCCD04]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNHAZZ07]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICUCNSXTY02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNGDFS03]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNZJNB02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNSCCD07]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNYNKM05]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNAHHF01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICUCNGZGY02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNSDJN02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNGXLZ01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNGSLZ03]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNBJBJ02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNSCYB01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALIMUTCNSCCD01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNSCYB02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALIMUTCNJXNC01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALIMUTCNGDFS01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALIMUTCNGDDG01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALIMUTCNHAZZ01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALIMUTCNLNDL01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALIMUTCNTJTJ01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALIMUTCNSXTY01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNSXTY02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALIMUTCNJXNC02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNQHHD01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALIMUTCNGZGY02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNYNKM06]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALIMUTCNHNLD01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALIMUTCNHESJZ02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNHELF01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICUCNSDQD03]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALIMUTCNJXNC03]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALIMUTCNHALY01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICUCNGXLZ01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICUCNGSLZ03]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALIMUTCNHESJZ03]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALIMUTCNGDDG02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALIMUTCNHBXY02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNLNDL04]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNNXZW03]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNSDQD03]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALIMUTCNFJQZ01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNLNDL05]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNSXTY03]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNNMHHHT02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALIMUTCNSCDY01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALIMUTCNHESJZ04]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALIMUTCNGDGZ01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALIMUTCNZJJX01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICUCNZJJH01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNLNSY01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALIMUTCNAHHF01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALIMUTCNHBWH02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNSDQD04]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICUCNSNXA02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALIMUTCNLNSY01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICUCNJLCC02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNJSNJ03]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALIMUTCNYNKM01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICUCNYNKM03]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNBJBJ03]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICUCNZJJH02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICUCNJSCZ01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICUCNLNSY01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALIMUTCNZJJH01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNLNDL06]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNSDYT01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNNXZW01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNGDGZ03]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNGZGY04]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNCQCQ02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNHLHEB02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNTJTJ02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNGZGY05]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICUCNCQCQ02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALIMUTCNAHHF02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALIMUTCNZJWZ01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICUCNSNXY01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNHAZZ08]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNSXTY02]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICMCNSNBJ01]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_ALICTCNGXNN03]: exports.VREGION.VREGION_CHINAEDGE,
    [exports.VDC.VDC_CDN]: exports.VREGION.VREGION_CHINACDN,
    [exports.VDC.VDC_GALINC2]: exports.VREGION.VREGION_ALIYUNNC2,
    [exports.VDC.VDC_BOE]: exports.VREGION.VREGION_CHINABOE,
    [exports.VDC.VDC_COF]: exports.VREGION.VREGION_CHINABOE,
    [exports.VDC.VDC_DEVBOX]: exports.VREGION.VREGION_CHINABOE,
    [exports.VDC.VDC_BOETEST]: exports.VREGION.VREGION_CHINABOE,
    [exports.VDC.VDC_BOE2]: exports.VREGION.VREGION_CHINABOE2,
    [exports.VDC.VDC_SINFBOE]: exports.VREGION.VREGION_CHINASINFBOE,
    [exports.VDC.VDC_DEVBOXI18N]: exports.VREGION.VREGION_USBOE,
    [exports.VDC.VDC_BOEI18N]: exports.VREGION.VREGION_USBOE,
    [exports.VDC.VDC_BOETTP]: exports.VREGION.VREGION_USBOE,
    [exports.VDC.VDC_BOEVPC2]: exports.VREGION.VREGION_USBOE,
    [exports.VDC.VDC_BOESG]: exports.VREGION.VREGION_SINGAPOREBOE,
    [exports.VDC.VDC_DEVBOXSG]: exports.VREGION.VREGION_SINGAPOREBOE,
    [exports.VDC.VDC_BOEINF]: exports.VREGION.VREGION_CHINAINFBOE,
    [exports.VDC.VDC_AUTODEPLOYTEST]: exports.VREGION.VREGION_CHINAINFBOE,
    [exports.VDC.VDC_AUTODEPLOYCICD]: exports.VREGION.VREGION_CHINAINFBOE,
    [exports.VDC.VDC_MALIVA]: exports.VREGION.VREGION_USEAST,
    [exports.VDC.VDC_USEAST1A]: exports.VREGION.VREGION_USEAST,
    [exports.VDC.VDC_USEAST3]: exports.VREGION.VREGION_USEAST,
    [exports.VDC.VDC_USEAST4]: exports.VREGION.VREGION_USEAST,
    [exports.VDC.VDC_GCPBR]: exports.VREGION.VREGION_USEAST,
    [exports.VDC.VDC_AWSVAGM]: exports.VREGION.VREGION_USEAST,
    [exports.VDC.VDC_AWSVAC]: exports.VREGION.VREGION_USEAST,
    [exports.VDC.VDC_GCPUSIAD]: exports.VREGION.VREGION_USEAST,
    [exports.VDC.VDC_ALIVA]: exports.VREGION.VREGION_USEAST,
    [exports.VDC.VDC_USEASTDT]: exports.VREGION.VREGION_USEAST,
    [exports.VDC.VDC_USEASTAZURE]: exports.VREGION.VREGION_USEAST,
    [exports.VDC.VDC_USTSENTRY]: exports.VREGION.VREGION_USEAST,
    [exports.VDC.VDC_USEAST6]: exports.VREGION.VREGION_USEAST,
    [exports.VDC.VDC_USEAST7]: exports.VREGION.VREGION_USEAST,
    [exports.VDC.VDC_USEAST10A]: exports.VREGION.VREGION_USEAST,
    [exports.VDC.VDC_USEASTIC]: exports.VREGION.VREGION_USEAST,
    [exports.VDC.VDC_USEAST2AZURE]: exports.VREGION.VREGION_USEAST,
    [exports.VDC.VDC_USEAST1B]: exports.VREGION.VREGION_USEAST,
    [exports.VDC.VDC_ALISG]: exports.VREGION.VREGION_SINGAPORECENTRAL,
    [exports.VDC.VDC_SG]: exports.VREGION.VREGION_SINGAPORECENTRAL,
    [exports.VDC.VDC_SG1]: exports.VREGION.VREGION_SINGAPORECENTRAL,
    [exports.VDC.VDC_SGDT]: exports.VREGION.VREGION_SINGAPORECENTRAL,
    [exports.VDC.VDC_SGAZURE]: exports.VREGION.VREGION_SINGAPORECENTRAL,
    [exports.VDC.VDC_SGISOLATION]: exports.VREGION.VREGION_SINGAPORECENTRAL,
    [exports.VDC.VDC_SG2]: exports.VREGION.VREGION_SINGAPORECENTRAL,
    [exports.VDC.VDC_MY]: exports.VREGION.VREGION_SINGAPORECENTRAL,
    [exports.VDC.VDC_MY2]: exports.VREGION.VREGION_SINGAPORECENTRAL,
    [exports.VDC.VDC_MY3]: exports.VREGION.VREGION_SINGAPORECENTRAL,
    [exports.VDC.VDC_SGSAAS1LARKIDC1]: exports.VREGION.VREGION_SINGAPORESAAS,
    [exports.VDC.VDC_SGSAAS1LARKIDC2]: exports.VREGION.VREGION_SINGAPORESAAS,
    [exports.VDC.VDC_SGSAAS1LARKIDC3]: exports.VREGION.VREGION_SINGAPORESAAS,
    [exports.VDC.VDC_SGCOMPLIANCE]: exports.VREGION.VREGION_SINGAPORECOMPLIANCE,
    [exports.VDC.VDC_MYA]: exports.VREGION.VREGION_ASIASOUTHEASTBD,
    [exports.VDC.VDC_MYB]: exports.VREGION.VREGION_ASIASOUTHEASTBD,
    [exports.VDC.VDC_SG3A]: exports.VREGION.VREGION_ASIASOUTHEASTBD,
    [exports.VDC.VDC_ID3B]: exports.VREGION.VREGION_ASIASOUTHEASTBD,
    [exports.VDC.VDC_HK1A]: exports.VREGION.VREGION_ASIASOUTHEASTBD,
    [exports.VDC.VDC_MYC]: exports.VREGION.VREGION_ASIASOUTHEASTBD,
    [exports.VDC.VDC_BDSGDT]: exports.VREGION.VREGION_ASIASOUTHEASTBD,
    [exports.VDC.VDC_MY5A]: exports.VREGION.VREGION_ASIASOUTHEASTBD,
    [exports.VDC.VDC_ID4B]: exports.VREGION.VREGION_ASIASOUTHEASTBD,
    [exports.VDC.VDC_ID4A]: exports.VREGION.VREGION_ASIASOUTHEASTBD,
    [exports.VDC.VDC_MYCISA]: exports.VREGION.VREGION_ASIACIS,
    [exports.VDC.VDC_MYCISB]: exports.VREGION.VREGION_ASIACIS,
    [exports.VDC.VDC_SGCISA]: exports.VREGION.VREGION_ASIACIS,
    [exports.VDC.VDC_JPSAAS]: exports.VREGION.VREGION_ASIASAAS,
    [exports.VDC.VDC_JPSAAS2A]: exports.VREGION.VREGION_ASIASAAS,
    [exports.VDC.VDC_AWSFR]: exports.VREGION.VREGION_EUROPECENTRAL,
    [exports.VDC.VDC_GCPPL]: exports.VREGION.VREGION_EUROPECENTRAL,
    [exports.VDC.VDC_FRACOMPLIANCE]: exports.VREGION.VREGION_EUROPECENTRAL,
    [exports.VDC.VDC_BE1A]: exports.VREGION.VREGION_EUROPECENTRAL,
    [exports.VDC.VDC_FR1A]: exports.VREGION.VREGION_EUROPECENTRAL,
    [exports.VDC.VDC_ID1A]: exports.VREGION.VREGION_IDCOMPLIANCE,
    [exports.VDC.VDC_ID2A]: exports.VREGION.VREGION_IDCOMPLIANCE2,
    [exports.VDC.VDC_USEAST11A]: exports.VREGION.VREGION_USCOMPLIANCE,
    [exports.VDC.VDC_USWEST1A]: exports.VREGION.VREGION_USWEST,
    [exports.VDC.VDC_CA]: exports.VREGION.VREGION_USWEST,
    [exports.VDC.VDC_AWSUSWEST2]: exports.VREGION.VREGION_USWEST,
    [exports.VDC.VDC_USWEST2]: exports.VREGION.VREGION_USWEST,
    [exports.VDC.VDC_USWEST3A]: exports.VREGION.VREGION_USWEST,
    [exports.VDC.VDC_USWEST3B]: exports.VREGION.VREGION_USWEST,
    [exports.VDC.VDC_USSW1B]: exports.VREGION.VREGION_USSOUTHWEST,
    [exports.VDC.VDC_USSW1A]: exports.VREGION.VREGION_USSOUTHWEST,
    [exports.VDC.VDC_GCPUSCBF]: exports.VREGION.VREGION_USCENTRAL,
    [exports.VDC.VDC_GCPUNUSE]: exports.VREGION.VREGION_USCENTRAL,
    [exports.VDC.VDC_USORDX]: exports.VREGION.VREGION_USCENTRAL,
    [exports.VDC.VDC_USCENTRAL1A]: exports.VREGION.VREGION_USCENTRAL,
    [exports.VDC.VDC_VA]: exports.VREGION.VREGION_USEE,
    [exports.VDC.VDC_USEAST2A]: exports.VREGION.VREGION_USEASTRED,
    [exports.VDC.VDC_USEAST2B]: exports.VREGION.VREGION_USEASTRED,
    [exports.VDC.VDC_USEAST9A]: exports.VREGION.VREGION_USEASTBD,
    [exports.VDC.VDC_USEAST12A]: exports.VREGION.VREGION_USEASTBD,
    [exports.VDC.VDC_USEAST13A]: exports.VREGION.VREGION_USEASTBD,
    [exports.VDC.VDC_USEAST14A]: exports.VREGION.VREGION_USEASTBD,
    [exports.VDC.VDC_GCPCA]: exports.VREGION.VREGION_USNORTHEAST,
    [exports.VDC.VDC_SGCOMM1]: exports.VREGION.VREGION_SINGAPORECOMMON,
    [exports.VDC.VDC_SGEA2]: exports.VREGION.VREGION_SINGAPOREEE,
    [exports.VDC.VDC_SGEA1]: exports.VREGION.VREGION_SINGAPOREEE,
    [exports.VDC.VDC_SGEE2]: exports.VREGION.VREGION_SINGAPOREEE,
    [exports.VDC.VDC_SGEE1]: exports.VREGION.VREGION_SINGAPOREEE,
    [exports.VDC.VDC_SGEE3]: exports.VREGION.VREGION_SINGAPOREEE,
    [exports.VDC.VDC_GALISG]: exports.VREGION.VREGION_SINGAPOREGAME,
    [exports.VDC.VDC_GCPIN]: exports.VREGION.VREGION_ASIASOUTH,
    [exports.VDC.VDC_AWSIN]: exports.VREGION.VREGION_ASIASOUTH,
    [exports.VDC.VDC_IND]: exports.VREGION.VREGION_ASIASOUTH,
    [exports.VDC.VDC_INDDP]: exports.VREGION.VREGION_ASIASOUTH,
    [exports.VDC.VDC_GCPINDEL]: exports.VREGION.VREGION_ASIASOUTH,
    [exports.VDC.VDC_JOHORTOBIAAS]: exports.VREGION.VREGION_ASIAENTERPRISE,
    [exports.VDC.VDC_HKTOBIAAS]: exports.VREGION.VREGION_ASIAENTERPRISE,
    [exports.VDC.VDC_SGTOBIAAS]: exports.VREGION.VREGION_ASIAENTERPRISE,
    [exports.VDC.VDC_JKTTOBIAAS]: exports.VREGION.VREGION_ASIAENTERPRISE,
    [exports.VDC.VDC_GCPJPOSA]: exports.VREGION.VREGION_ASIANORTHEAST,
    [exports.VDC.VDC_GCPJPTKY]: exports.VREGION.VREGION_ASIANORTHEAST,
    [exports.VDC.VDC_AWSJP]: exports.VREGION.VREGION_ASIANORTHEAST,
    [exports.VDC.VDC_GCPSG]: exports.VREGION.VREGION_ASIASOUTHEAST,
    [exports.VDC.VDC_AWSSG]: exports.VREGION.VREGION_ASIASOUTHEAST,
    [exports.VDC.VDC_GCPID]: exports.VREGION.VREGION_ASIASOUTHEAST,
    [exports.VDC.VDC_AWSIDHLP]: exports.VREGION.VREGION_ASIASOUTHEAST,
    [exports.VDC.VDC_AWSKR]: exports.VREGION.VREGION_ASIASOUTHEAST,
    [exports.VDC.VDC_ASID1A]: exports.VREGION.VREGION_ASIASOUTHEAST,
    [exports.VDC.VDC_ASID2A]: exports.VREGION.VREGION_ASIASOUTHEAST,
    [exports.VDC.VDC_AWSDEGM]: exports.VREGION.VREGION_I18NGAME,
    [exports.VDC.VDC_AWSJPGM]: exports.VREGION.VREGION_I18NGAME,
    [exports.VDC.VDC_AWSSGGM]: exports.VREGION.VREGION_I18NGAME,
    [exports.VDC.VDC_PROGAME]: exports.VREGION.VREGION_I18NGAME,
    [exports.VDC.VDC_GCPKR]: exports.VREGION.VREGION_I18NGAME,
    [exports.VDC.VDC_BGE]: exports.VREGION.VREGION_I18NBGE,
    [exports.VDC.VDC_GCPSG2]: exports.VREGION.VREGION_I18NBGE,
    [exports.VDC.VDC_GCPAU]: exports.VREGION.VREGION_AUSTRALIASOUTHEAST,
    [exports.VDC.VDC_SYD1A]: exports.VREGION.VREGION_AUSTRALIASOUTHEAST,
    [exports.VDC.VDC_YCRU]: exports.VREGION.VREGION_EASTERNEUROTT,
    [exports.VDC.VDC_DUBCOMPLIANCE]: exports.VREGION.VREGION_EUROPECOMPLIANCE,
    [exports.VDC.VDC_FR]: exports.VREGION.VREGION_EUROPECOMPLIANCE,
    [exports.VDC.VDC_KSRU]: exports.VREGION.VREGION_EUROPEEAST,
    [exports.VDC.VDC_GCPFI]: exports.VREGION.VREGION_EUROPENORTH,
    [exports.VDC.VDC_GCPBE]: exports.VREGION.VREGION_EUROPEWEST,
    [exports.VDC.VDC_GCPGB]: exports.VREGION.VREGION_EUROPEWEST,
    [exports.VDC.VDC_GCPDE]: exports.VREGION.VREGION_EUROPEWEST,
    [exports.VDC.VDC_GCPNL]: exports.VREGION.VREGION_EUROPEWEST,
    [exports.VDC.VDC_GCPCH]: exports.VREGION.VREGION_EUROPEWEST,
    [exports.VDC.VDC_AWSUKLD]: exports.VREGION.VREGION_EUROPEWEST,
    [exports.VDC.VDC_AWSBH]: exports.VREGION.VREGION_MIDDLEEASTSOUTH,
    [exports.VDC.VDC_RTCSA]: exports.VREGION.VREGION_MIDDLEEASTSOUTH,
    [exports.VDC.VDC_AWSBR]: exports.VREGION.VREGION_SOUTHAMERICAEAST,
    [exports.VDC.VDC_AWSSFCPT]: exports.VREGION.VREGION_AFRICASOUTH,
    [exports.VDC.VDC_USEAST5]: exports.VREGION.VREGION_USTTP,
    [exports.VDC.VDC_USEAST8]: exports.VREGION.VREGION_USTTP2,
    [exports.VDC.VDC_IE]: exports.VREGION.VREGION_EUTTP,
    [exports.VDC.VDC_DEDT]: exports.VREGION.VREGION_EUTTP,
    [exports.VDC.VDC_IEDT]: exports.VREGION.VREGION_EUTTP,
    [exports.VDC.VDC_EUIE1A]: exports.VREGION.VREGION_EUTTP,
    [exports.VDC.VDC_IEAZURE]: exports.VREGION.VREGION_EUTTP,
    [exports.VDC.VDC_NO]: exports.VREGION.VREGION_EUTTP2,
    [exports.VDC.VDC_NO1A]: exports.VREGION.VREGION_EUTTP2,
    [exports.VDC.VDC_NOAZURE]: exports.VREGION.VREGION_EUTTP2,
    [exports.VDC.VDC_DE]: exports.VREGION.VREGION_EUCOMPLIANCE,
    [exports.VDC.VDC_IE2]: exports.VREGION.VREGION_EUCOMPLIANCE2,
    [exports.VDC.VDC_JPSOUTH3A]: exports.VREGION.VREGION_ASIANORTHEASTBD,
    [exports.VDC.VDC_DY]: exports.VREGION.VREGION_CHINANORTH3,
    [exports.VDC.VDC_YX]: exports.VREGION.VREGION_CHINANORTH3,
    [exports.VDC.VDC_ZC]: exports.VREGION.VREGION_CHINANORTH3,
    [exports.VDC.VDC_VATOBIAAS]: exports.VREGION.VREGION_USENTERPRISE,
    [exports.VDC.VDC_HLHKPAY]: exports.VREGION.VREGION_CHINAHKPAY,
    [exports.VDC.VDC_LFHKPAY]: exports.VREGION.VREGION_CHINAHKPAY,
    [exports.VDC.VDC_IE3A]: exports.VREGION.VREGION_EUROPEWESTBD,
    [exports.VDC.VDC_BDDEDT]: exports.VREGION.VREGION_EUROPEWESTBD,
    [exports.VDC.VDC_BE2A]: exports.VREGION.VREGION_EUROPEWESTBD,
    [exports.VDC.VDC_DE2A]: exports.VREGION.VREGION_EUROPEWESTBD,
    [exports.VDC.VDC_MY4A]: exports.VREGION.VREGION_MYCOMPLIANCE,
    [exports.VDC.VDC_WLBY]: exports.VREGION.VREGION_CHINANORTH5,
    [exports.VDC.VDC_GL2]: exports.VREGION.VREGION_CHINANORTH6,
    [exports.VDC.VDC_XH]: exports.VREGION.VREGION_CHINANORTH6,
    [exports.VDC.VDC_ZB]: exports.VREGION.VREGION_CHINANORTH6,
    [exports.VDC.VDC_HLTOBIAASBOE]: exports.VREGION.VREGION_CHINAENTERPRISEBOE,
    [exports.VDC.VDC_BGE2]: exports.VREGION.VREGION_EUBGE,
    [exports.VDC.VDC_BGEEU]: exports.VREGION.VREGION_EUBGE,
    [exports.VDC.VDC_SYD2A]: exports.VREGION.VREGION_AUSTRALIASOUTHEASTBD,
    [exports.VDC.VDC_USEAST15A]: exports.VREGION.VREGION_USTTP3,
    [exports.VDC.VDC_TH1A]: exports.VREGION.VREGION_THCENTRAL,
    [exports.VDC.VDC_TH1B]: exports.VREGION.VREGION_THCENTRAL,
};
