// @universe-design/styles/esm/preset-dark.atom.js
// autogen by ../scripts/gen-ts-from-json.ts
export const AI_cursor = 'rgb(94, 104, 232)';
export const AI_dynamic_line = 'conic-gradient(from 45deg, rgb(164, 61, 164) 0%, rgb(94, 104, 232) 33.33333333333333%, rgb(94, 104, 232) 66.66666666666666%, rgb(164, 61, 164) 100%)';
export const AI_icon_gradient = 'linear-gradient(to right, rgb(123, 131, 247) 0%, rgb(140, 85, 236) 100%)';
export const AI_icon_in_button = 'rgb(123, 131, 247)';
export const AI_loading = 'linear-gradient(to right, rgb(123, 131, 247) 0%, rgb(164, 61, 164) 100%)';
export const AI_primary_content_default = 'linear-gradient(to right, rgb(123, 131, 247) 0%, rgb(181, 74, 181) 100%)';
export const AI_primary_content_hover = 'linear-gradient(to right, rgb(94, 104, 232) 0%, rgb(164, 61, 164) 100%)';
export const AI_primary_content_loading = 'linear-gradient(to right, rgba(123, 131, 247, 0.6) 0%, rgba(181, 74, 181, 0.6) 100%)';
export const AI_primary_content_pressed = 'linear-gradient(to right, rgb(148, 153, 247) 0%, rgb(214, 97, 214) 100%)';
export const AI_primary_fill_default = 'linear-gradient(to right, rgb(94, 104, 232) 0%, rgb(139, 55, 139) 100%)';
export const AI_primary_fill_hover = 'linear-gradient(to right, rgb(81, 90, 214) 0%, rgb(113, 40, 113) 100%)';
export const AI_primary_fill_loading = 'linear-gradient(to right, rgba(94, 104, 232, 0.6) 0%, rgba(139, 55, 139, 0.6) 100%)';
export const AI_primary_fill_pressed = 'linear-gradient(to right, rgb(123, 131, 247) 0%, rgb(164, 61, 164) 100%)';
export const AI_primary_fill_solid_01 = 'linear-gradient(to right, rgb(30, 32, 74) 0%, rgb(59, 21, 59) 100%)';
export const AI_primary_fill_solid_02 = 'linear-gradient(to right, rgb(42, 45, 105) 0%, rgb(84, 24, 84) 100%)';
export const AI_primary_fill_solid_03 = 'linear-gradient(to right, rgb(55, 61, 144) 0%, rgb(113, 40, 113) 100%)';
export const AI_primary_fill_transparent_01 = 'linear-gradient(to right, rgba(94, 104, 232, 0.15) 0%, rgba(164, 61, 164, 0.15) 100%)';
export const AI_primary_fill_transparent_02 = 'linear-gradient(to right, rgba(94, 104, 232, 0.25) 0%, rgba(164, 61, 164, 0.25) 100%)';
export const AI_primary_fill_transparent_03 = 'linear-gradient(to right, rgba(94, 104, 232, 0.35) 0%, rgba(164, 61, 164, 0.35) 100%)';
export const AI_sendicon = 'linear-gradient(to right, rgb(94, 104, 232) 0%, rgb(140, 85, 236) 100%)';
export const Aurora_bg_main = 'rgba(16, 153, 204, 0.2)';
export const Aurora_bg_reflection = 'rgba(165, 117, 250, 0.15)';
export const Aurora_bg_subordinate = 'rgba(16, 153, 204, 0.2)';
export const B100 = '#173166';
export const B200 = '#194294';
export const B300 = '#2655b6';
export const B350 = '#275fce';
export const B400 = '#3370eb';
export const B50 = '#152340';
export const B500 = '#4c88ff';
export const B600 = '#75a4ff';
export const B700 = '#8fb4ff';
export const B800 = '#bdd2ff';
export const B900 = '#e0e9ff';
export const C100 = '#591c3f';
export const C200 = '#782b57';
export const C300 = '#94386c';
export const C350 = '#ab417d';
export const C400 = '#c24a8e';
export const C50 = '#3a182b';
export const C500 = '#db5ea4';
export const C600 = '#ed77ba';
export const C700 = '#fc94cf';
export const C800 = '#ffc2e5';
export const C900 = '#ffe0f2';
export const G100 = '#173b12';
export const G200 = '#21511a';
export const G300 = '#296621';
export const G350 = '#2f7526';
export const G400 = '#35872a';
export const G50 = '#0e2b0a';
export const G500 = '#419e34';
export const G600 = '#51ba43';
export const G700 = '#69cc5c';
export const G800 = '#99e490';
export const G900 = '#cbf5c6';
export const I100 = '#2a2d69';
export const I200 = '#373d90';
export const I300 = '#474fb8';
export const I350 = '#515ad6';
export const I400 = '#5e68e8';
export const I50 = '#1e204a';
export const I500 = '#7b83f7';
export const I600 = '#9499f7';
export const I700 = '#aaaff8';
export const I800 = '#cdd0f9';
export const I900 = '#e7e9fe';
export const L100 = '#303804';
export const L200 = '#404c06';
export const L300 = '#53610a';
export const L350 = '#5f700a';
export const L400 = '#6b7f05';
export const L50 = '#212702';
export const L500 = '#82990a';
export const L600 = '#93af04';
export const L700 = '#a6c313';
export const L800 = '#c2e12d';
export const L900 = '#e3f391';
export const N00 = '#0a0a0a';
export const N00_10 = 'rgba(10, 10, 10, 0.1)';
export const N00_15 = 'rgba(10, 10, 10, 0.15)';
export const N00_20 = 'rgba(10, 10, 10, 0.2)';
export const N00_30 = 'rgba(10, 10, 10, 0.3)';
export const N00_40 = 'rgba(10, 10, 10, 0.4)';
export const N00_5 = 'rgba(10, 10, 10, 0.05)';
export const N00_50 = 'rgba(10, 10, 10, 0.5)';
export const N00_60 = 'rgba(10, 10, 10, 0.6)';
export const N00_70 = 'rgba(10, 10, 10, 0.7)';
export const N00_80 = 'rgba(10, 10, 10, 0.8)';
export const N00_90 = 'rgba(10, 10, 10, 0.9)';
export const N100 = '#292929';
export const N1000 = '#ffffff';
export const N200 = '#373737';
export const N300 = '#434343';
export const N350 = '#505050';
export const N400 = '#5f5f5f';
export const N50 = '#1a1a1a';
export const N500 = '#757575';
export const N600 = '#a6a6a6';
export const N650 = '#cfcfcf';
export const N700 = '#e0e0e0';
export const N800 = '#e8e8e8';
export const N900 = '#ebebeb';
export const N900_10 = 'rgba(235, 235, 235, 0.1)';
export const N900_15 = 'rgba(235, 235, 235, 0.15)';
export const N900_20 = 'rgba(235, 235, 235, 0.2)';
export const N900_30 = 'rgba(235, 235, 235, 0.3)';
export const N900_40 = 'rgba(235, 235, 235, 0.4)';
export const N900_5 = 'rgba(235, 235, 235, 0.05)';
export const N900_50 = 'rgba(235, 235, 235, 0.5)';
export const N900_60 = 'rgba(235, 235, 235, 0.6)';
export const N900_70 = 'rgba(235, 235, 235, 0.7)';
export const N900_80 = 'rgba(235, 235, 235, 0.8)';
export const N900_90 = 'rgba(235, 235, 235, 0.9)';
export const N950 = '#f8f8f8';
export const O100 = '#4a2b10';
export const O200 = '#683a12';
export const O300 = '#8a4a19';
export const O350 = '#a15317';
export const O400 = '#b85e1a';
export const O50 = '#33210b';
export const O500 = '#db7018';
export const O600 = '#f3871b';
export const O700 = '#f89e44';
export const O800 = '#fec88b';
export const O900 = '#fee7cd';
export const P100 = '#3f2073';
export const P200 = '#5529a3';
export const P300 = '#6c39c6';
export const P350 = '#7c4ad4';
export const P400 = '#8c55ec';
export const P50 = '#2b194a';
export const P500 = '#a575fa';
export const P600 = '#b88ffe';
export const P700 = '#c5a3ff';
export const P800 = '#dbc8fd';
export const P900 = '#efe5ff';
export const R100 = '#591f1d';
export const R200 = '#7b2524';
export const R300 = '#a03331';
export const R350 = '#b33a37';
export const R400 = '#d14642';
export const R50 = '#3d1a19';
export const R500 = '#f05b56';
export const R600 = '#f6827e';
export const R700 = '#f89e9b';
export const R800 = '#fdc6c4';
export const R900 = '#fee3e2';
export const S100 = '#38320c';
export const S200 = '#574d01';
export const S300 = '#7a6a01';
export const S350 = '#9c8702';
export const S400 = '#c9b218';
export const S50 = '#29250a';
export const S500 = '#e5cd17';
export const S600 = '#f5df36';
export const S700 = '#f7e663';
export const S800 = '#faed7a';
export const S900 = '#fff59e';
export const T100 = '#173b36';
export const T200 = '#1d4e47';
export const T300 = '#25665e';
export const T350 = '#227369';
export const T400 = '#198578';
export const T50 = '#132926';
export const T500 = '#1fa18f';
export const T600 = '#1ab7a1';
export const T700 = '#17cfb5';
export const T800 = '#65e7d5';
export const T900 = '#b7f7ef';
export const V100 = '#541854';
export const V200 = '#712871';
export const V300 = '#8b378b';
export const V350 = '#a43da4';
export const V400 = '#b54ab5';
export const V50 = '#3b153b';
export const V500 = '#d661d6';
export const V600 = '#e17fe1';
export const V700 = '#e99be9';
export const V800 = '#f4c3f4';
export const V900 = '#fcdffc';
export const Vnext_bg_content_header = 'rgb(34, 34, 34)';
export const Vnext_bg_gradient = 'linear-gradient(180deg, rgba(26, 26, 26, 0.05) 0%, rgba(10, 10, 10, 0.05) 100%)';
export const Vnext_bg_sub_navigation = 'rgba(30, 30, 30, 0.5)';
export const Vnext_bg_thirdly_navigation = 'rgba(26, 26, 26, 0.7)';
export const W100 = '#103647';
export const W200 = '#164359';
export const W300 = '#135a78';
export const W350 = '#146c91';
export const W400 = '#1a7fab';
export const W50 = '#152830';
export const W500 = '#1099cc';
export const W600 = '#25b2e5';
export const W700 = '#51c3f0';
export const W800 = '#89dffe';
export const W900 = '#c7f0ff';
export const Y100 = '#473409';
export const Y200 = '#63470f';
export const Y300 = '#8a6419';
export const Y350 = '#ad7d15';
export const Y400 = '#d49b0b';
export const Y50 = '#30250a';
export const Y500 = '#f0b622';
export const Y600 = '#fbcb46';
export const Y700 = '#fcd456';
export const Y800 = '#ffde75';
export const Y900 = '#ffeaa3';
export const bg_base = 'rgb(10, 10, 10)';
export const bg_body = 'rgb(26, 26, 26)';
export const bg_body_overlay = 'rgb(41, 41, 41)';
export const bg_content_base = 'rgb(18, 18, 18)';
export const bg_filler = 'rgb(55, 55, 55)';
export const bg_float = 'rgb(41, 41, 41)';
export const bg_float_base = 'rgb(26, 26, 26)';
export const bg_float_overlay = 'rgb(55, 55, 55)';
export const bg_float_push = 'rgba(41, 41, 41, 0.8)';
export const bg_mask = 'rgba(0, 0, 0, 0.6)';
export const bg_pricolor = 'rgb(51, 112, 235)';
export const bg_sub_navigation = 'rgb(38, 38, 38)';
export const bg_text_selection = 'rgba(117, 164, 255, 0.3)';
export const bg_tips = 'rgb(80, 80, 80)';
export const colorful_blue = 'rgb(76, 136, 255)';
export const colorful_carmine = 'rgb(194, 74, 142)';
export const colorful_green = 'rgb(65, 158, 52)';
export const colorful_indigo = 'rgb(94, 104, 232)';
export const colorful_lime = 'rgb(147, 175, 4)';
export const colorful_neutral = 'rgb(117, 117, 117)';
export const colorful_orange = 'rgb(243, 135, 27)';
export const colorful_purple = 'rgb(140, 85, 236)';
export const colorful_red = 'rgb(240, 91, 86)';
export const colorful_sunflower = 'rgb(229, 205, 23)';
export const colorful_turquoise = 'rgb(31, 161, 143)';
export const colorful_violet = 'rgb(181, 74, 181)';
export const colorful_wathet = 'rgb(16, 153, 204)';
export const colorful_yellow = 'rgb(240, 182, 34)';
export const fill_active = 'rgba(76, 136, 255, 0.2)';
export const fill_disabled = 'rgb(95, 95, 95)';
export const fill_focus = 'rgba(235, 235, 235, 0.12)';
export const fill_hover = 'rgba(235, 235, 235, 0.08)';
export const fill_img_mask = 'rgba(0, 0, 0, 0.12)';
export const fill_loading_mask = 'rgba(10, 10, 10, 0.6)';
export const fill_pressed = 'rgba(235, 235, 235, 0.12)';
export const fill_selected = 'rgba(76, 136, 255, 0.15)';
export const fill_tag = 'rgba(235, 235, 235, 0.1)';
export const function_danger_100 = 'rgb(89, 31, 29)';
export const function_danger_200 = 'rgb(123, 37, 36)';
export const function_danger_300 = 'rgb(160, 51, 49)';
export const function_danger_350 = 'rgb(179, 58, 55)';
export const function_danger_400 = 'rgb(209, 70, 66)';
export const function_danger_50 = 'rgb(61, 26, 25)';
export const function_danger_500 = 'rgb(240, 91, 86)';
export const function_danger_600 = 'rgb(246, 130, 126)';
export const function_danger_700 = 'rgb(248, 158, 155)';
export const function_danger_800 = 'rgb(253, 198, 196)';
export const function_danger_900 = 'rgb(254, 227, 226)';
export const function_danger_content_default = 'rgb(240, 91, 86)';
export const function_danger_content_hover = 'rgb(209, 70, 66)';
export const function_danger_content_loading = 'rgb(160, 51, 49)';
export const function_danger_content_pressed = 'rgb(246, 130, 126)';
export const function_danger_fill_default = 'rgb(209, 70, 66)';
export const function_danger_fill_hover = 'rgb(179, 58, 55)';
export const function_danger_fill_loading = 'rgb(160, 51, 49)';
export const function_danger_fill_pressed = 'rgb(240, 91, 86)';
export const function_danger_fill_solid_01 = 'rgb(61, 26, 25)';
export const function_danger_fill_solid_02 = 'rgb(89, 31, 29)';
export const function_danger_fill_solid_03 = 'rgb(123, 37, 36)';
export const function_danger_fill_transparent_01 = 'rgba(240, 91, 86, 0.15)';
export const function_danger_fill_transparent_02 = 'rgba(240, 91, 86, 0.2)';
export const function_danger_fill_transparent_03 = 'rgba(240, 91, 86, 0.3)';
export const function_danger_on_danger_fill = 'rgb(255, 255, 255)';
export const function_info_100 = 'rgb(23, 49, 102)';
export const function_info_200 = 'rgb(25, 66, 148)';
export const function_info_300 = 'rgb(38, 85, 182)';
export const function_info_350 = 'rgb(39, 95, 206)';
export const function_info_400 = 'rgb(51, 112, 235)';
export const function_info_50 = 'rgb(21, 35, 64)';
export const function_info_500 = 'rgb(76, 136, 255)';
export const function_info_600 = 'rgb(117, 164, 255)';
export const function_info_700 = 'rgb(143, 180, 255)';
export const function_info_800 = 'rgb(189, 210, 255)';
export const function_info_900 = 'rgb(224, 233, 255)';
export const function_info_content_default = 'rgb(76, 136, 255)';
export const function_info_content_hover = 'rgb(51, 112, 235)';
export const function_info_content_loading = 'rgb(38, 85, 182)';
export const function_info_content_pressed = 'rgb(117, 164, 255)';
export const function_info_fill_default = 'rgb(51, 112, 235)';
export const function_info_fill_hover = 'rgb(39, 95, 206)';
export const function_info_fill_loading = 'rgb(38, 85, 182)';
export const function_info_fill_pressed = 'rgb(76, 136, 255)';
export const function_info_fill_solid_01 = 'rgb(21, 35, 64)';
export const function_info_fill_solid_02 = 'rgb(23, 49, 102)';
export const function_info_fill_solid_03 = 'rgb(25, 66, 148)';
export const function_info_fill_transparent_01 = 'rgba(76, 136, 255, 0.15)';
export const function_info_fill_transparent_02 = 'rgba(76, 136, 255, 0.2)';
export const function_info_fill_transparent_03 = 'rgba(76, 136, 255, 0.3)';
export const function_info_on_info_fill = 'rgb(255, 255, 255)';
export const function_success_100 = 'rgb(23, 59, 18)';
export const function_success_200 = 'rgb(33, 81, 26)';
export const function_success_300 = 'rgb(41, 102, 33)';
export const function_success_350 = 'rgb(47, 117, 38)';
export const function_success_400 = 'rgb(53, 135, 42)';
export const function_success_50 = 'rgb(14, 43, 10)';
export const function_success_500 = 'rgb(65, 158, 52)';
export const function_success_600 = 'rgb(81, 186, 67)';
export const function_success_700 = 'rgb(105, 204, 92)';
export const function_success_800 = 'rgb(153, 228, 144)';
export const function_success_900 = 'rgb(203, 245, 198)';
export const function_success_content_default = 'rgb(81, 186, 67)';
export const function_success_content_hover = 'rgb(65, 158, 52)';
export const function_success_content_loading = 'rgb(41, 102, 33)';
export const function_success_content_pressed = 'rgb(105, 204, 92)';
export const function_success_fill_default = 'rgb(65, 158, 52)';
export const function_success_fill_hover = 'rgb(53, 135, 42)';
export const function_success_fill_loading = 'rgb(41, 102, 33)';
export const function_success_fill_pressed = 'rgb(81, 186, 67)';
export const function_success_fill_solid_01 = 'rgb(14, 43, 10)';
export const function_success_fill_solid_02 = 'rgb(23, 59, 18)';
export const function_success_fill_solid_03 = 'rgb(33, 81, 26)';
export const function_success_fill_transparent_01 = 'rgba(65, 158, 52, 0.15)';
export const function_success_fill_transparent_02 = 'rgba(65, 158, 52, 0.2)';
export const function_success_fill_transparent_03 = 'rgba(65, 158, 52, 0.3)';
export const function_success_on_success_fill = 'rgb(255, 255, 255)';
export const function_warning_100 = 'rgb(74, 43, 16)';
export const function_warning_200 = 'rgb(104, 58, 18)';
export const function_warning_300 = 'rgb(138, 74, 25)';
export const function_warning_350 = 'rgb(161, 83, 23)';
export const function_warning_400 = 'rgb(184, 94, 26)';
export const function_warning_50 = 'rgb(51, 33, 11)';
export const function_warning_500 = 'rgb(219, 112, 24)';
export const function_warning_600 = 'rgb(243, 135, 27)';
export const function_warning_700 = 'rgb(248, 158, 68)';
export const function_warning_800 = 'rgb(254, 200, 139)';
export const function_warning_900 = 'rgb(254, 231, 205)';
export const function_warning_content_default = 'rgb(243, 135, 27)';
export const function_warning_content_hover = 'rgb(219, 112, 24)';
export const function_warning_content_loading = 'rgb(138, 74, 25)';
export const function_warning_content_pressed = 'rgb(248, 158, 68)';
export const function_warning_fill_default = 'rgb(243, 135, 27)';
export const function_warning_fill_hover = 'rgb(219, 112, 24)';
export const function_warning_fill_loading = 'rgb(138, 74, 25)';
export const function_warning_fill_pressed = 'rgb(248, 158, 68)';
export const function_warning_fill_solid_01 = 'rgb(51, 33, 11)';
export const function_warning_fill_solid_02 = 'rgb(74, 43, 16)';
export const function_warning_fill_solid_03 = 'rgb(104, 58, 18)';
export const function_warning_fill_transparent_01 = 'rgba(243, 135, 27, 0.15)';
export const function_warning_fill_transparent_02 = 'rgba(243, 135, 27, 0.2)';
export const function_warning_fill_transparent_03 = 'rgba(243, 135, 27, 0.3)';
export const function_warning_on_warning_fill = 'rgb(255, 255, 255)';
export const gradient_blue = 'linear-gradient(90deg, rgb(51, 112, 235) 0%, rgb(39, 95, 206) 100%)';
export const gradient_carmine = 'linear-gradient(90deg, rgb(219, 94, 164) 0%, rgb(194, 74, 142) 100%)';
export const gradient_green = 'linear-gradient(90deg, rgb(65, 158, 52) 0%, rgb(53, 135, 42) 100%)';
export const gradient_indigo = 'linear-gradient(90deg, rgb(94, 104, 232) 0%, rgb(81, 90, 214) 100%)';
export const gradient_lime = 'linear-gradient(90deg, rgb(147, 175, 4) 0%, rgb(130, 153, 10) 100%)';
export const gradient_orange = 'linear-gradient(90deg, rgb(243, 135, 27) 0%, rgb(219, 112, 24) 100%)';
export const gradient_purple = 'linear-gradient(90deg, rgb(140, 85, 236) 0%, rgb(124, 74, 212) 100%)';
export const gradient_red = 'linear-gradient(90deg, rgb(240, 91, 86) 0%, rgb(209, 70, 66) 100%)';
export const gradient_turquoise = 'linear-gradient(90deg, rgb(26, 183, 161) 0%, rgb(31, 161, 143) 100%)';
export const gradient_violet = 'linear-gradient(90deg, rgb(214, 97, 214) 0%, rgb(181, 74, 181) 100%)';
export const gradient_wathet = 'linear-gradient(90deg, rgb(16, 153, 204) 0%, rgb(26, 127, 171) 100%)';
export const gradient_yellow = 'linear-gradient(90deg, rgb(240, 182, 34) 0%, rgb(212, 155, 11) 100%)';
export const icon_disabled = 'rgb(95, 95, 95)';
export const icon_n1 = 'rgb(232, 232, 232)';
export const icon_n2 = 'rgb(166, 166, 166)';
export const icon_n3 = 'rgb(117, 117, 117)';
export const illustration_blue_a = 'rgb(224, 224, 224)';
export const illustration_blue_b = 'rgba(255, 255, 255, 0.4)';
export const illustration_blue_c = 'rgb(95, 95, 95)';
export const illustration_blue_d = 'rgb(38, 85, 182)';
export const illustration_blue_e = 'rgb(76, 136, 255)';
export const illustration_neutral_a = 'rgb(235, 235, 235)';
export const illustration_neutral_b = 'rgb(55, 55, 55)';
export const illustration_neutral_c = 'rgba(255, 255, 255, 0.3)';
export const illustration_neutral_d = 'rgba(255, 255, 255, 0.4)';
export const illustration_orange = 'rgb(243, 135, 27)';
export const illustration_red = 'rgb(240, 91, 86)';
export const illustration_static_black = 'rgb(0, 0, 0)';
export const illustration_static_white = 'rgb(255, 255, 255)';
export const illustration_turquoise = 'rgb(23, 207, 181)';
export const illustration_wathet = 'rgb(81, 195, 240)';
export const illustration_yellow = 'rgb(240, 182, 34)';
export const inverse_primary_content_default = 'rgb(117, 164, 255)';
export const inverse_primary_content_hover = 'rgb(143, 180, 255)';
export const inverse_primary_content_pressed = 'rgb(76, 136, 255)';
export const line_border_card = 'rgba(235, 235, 235, 0.15)';
export const line_border_component = 'rgb(80, 80, 80)';
export const line_divider_default = 'rgba(207, 207, 207, 0.15)';
export const line_divider_module = 'rgb(0, 0, 0)';
export const primary_content_default = 'rgb(76, 136, 255)';
export const primary_content_hover = 'rgb(51, 112, 235)';
export const primary_content_loading = 'rgb(38, 85, 182)';
export const primary_content_pressed = 'rgb(117, 164, 255)';
export const primary_fill_default = 'rgb(51, 112, 235)';
export const primary_fill_hover = 'rgb(39, 95, 206)';
export const primary_fill_loading = 'rgb(38, 85, 182)';
export const primary_fill_pressed = 'rgb(76, 136, 255)';
export const primary_fill_solid_01 = 'rgb(21, 35, 64)';
export const primary_fill_solid_02 = 'rgb(23, 49, 102)';
export const primary_fill_solid_03 = 'rgb(25, 66, 148)';
export const primary_fill_transparent_01 = 'rgba(76, 136, 255, 0.15)';
export const primary_fill_transparent_02 = 'rgba(76, 136, 255, 0.2)';
export const primary_fill_transparent_03 = 'rgba(76, 136, 255, 0.3)';
export const primary_on_primary_fill = 'rgb(255, 255, 255)';
export const primary_pri_100 = 'rgb(23, 49, 102)';
export const primary_pri_200 = 'rgb(25, 66, 148)';
export const primary_pri_300 = 'rgb(38, 85, 182)';
export const primary_pri_350 = 'rgb(39, 95, 206)';
export const primary_pri_400 = 'rgb(51, 112, 235)';
export const primary_pri_50 = 'rgb(21, 35, 64)';
export const primary_pri_500 = 'rgb(76, 136, 255)';
export const primary_pri_600 = 'rgb(117, 164, 255)';
export const primary_pri_700 = 'rgb(143, 180, 255)';
export const primary_pri_800 = 'rgb(189, 210, 255)';
export const primary_pri_900 = 'rgb(224, 233, 255)';
export const shadow_default = 'rgb(0, 0, 0)';
export const shadow_default_lg = 'rgba(0, 0, 0, 0.24)';
export const shadow_default_md = 'rgba(0, 0, 0, 0.28)';
export const shadow_default_sm = 'rgba(0, 0, 0, 0.32)';
export const shadow_lg_down = '0px 12px 48px 0px rgba(0, 0, 0, 0.24)';
export const shadow_lg_left = '-12px 0px 48px 0px rgba(0, 0, 0, 0.24)';
export const shadow_lg_right = '12px 0px 48px 0px rgba(0, 0, 0, 0.24)';
export const shadow_lg_up = '0px -12px 48px 0px rgba(0, 0, 0, 0.24)';
export const shadow_md_down = '0px 8px 16px 0px rgba(0, 0, 0, 0.28)';
export const shadow_md_left = '-8px 0px 16px 0px rgba(0, 0, 0, 0.28)';
export const shadow_md_right = '8px 0px 16px 0px rgba(0, 0, 0, 0.28)';
export const shadow_md_up = '0px -8px 16px 0px rgba(0, 0, 0, 0.28)';
export const shadow_pri = 'rgb(36, 91, 219)';
export const shadow_pri_lg = 'rgba(25, 66, 148, 0.48)';
export const shadow_pri_lg_down = '0px 12px 48px 0px rgba(25, 66, 148, 0.48)';
export const shadow_pri_md = 'rgba(25, 66, 148, 0.48)';
export const shadow_pri_md_down = '0px 8px 16px 0px rgba(25, 66, 148, 0.48)';
export const shadow_s1_down = '0px 1px 2px -2px rgba(0, 0, 0, 0.02), 0px 2px 4px 0px rgba(0, 0, 0, 0.02), 0px 2px 8px 2px rgba(0, 0, 0, 0.02)';
export const shadow_s1_down_pri = '0px 1px 2px -2px rgba(36, 91, 219, 0.02), 0px 2px 4px 0px rgba(36, 91, 219, 0.02), 0px 2px 8px 2px rgba(36, 91, 219, 0.02)';
export const shadow_s1_left = '-1px 0px 2px -2px rgba(0, 0, 0, 0.02), -2px 0px 4px 0px rgba(0, 0, 0, 0.02), -2px 0px 8px 2px rgba(0, 0, 0, 0.02)';
export const shadow_s1_right = '1px 0px 2px -2px rgba(0, 0, 0, 0.02), 2px 0px 4px 0px rgba(0, 0, 0, 0.02), 2px 0px 8px 2px rgba(0, 0, 0, 0.02)';
export const shadow_s1_up = '0px -1px 2px -2px rgba(0, 0, 0, 0.02), 0px -2px 4px 0px rgba(0, 0, 0, 0.02), 0px -2px 8px 2px rgba(0, 0, 0, 0.02)';
export const shadow_s2_down = '0px 4px 16px 4px rgba(0, 0, 0, 0.03), 0px 4px 8px 0px rgba(0, 0, 0, 0.02), 0px 2px 4px -4px rgba(0, 0, 0, 0.02)';
export const shadow_s2_down_pri = '0px 4px 16px 4px rgba(36, 91, 219, 0.03), 0px 4px 8px 0px rgba(36, 91, 219, 0.02), 0px 2px 4px -4px rgba(36, 91, 219, 0.02)';
export const shadow_s2_left = '-4px 0px 16px 4px rgba(0, 0, 0, 0.03), -4px 0px 8px 0px rgba(0, 0, 0, 0.02), -2px 0px 4px -4px rgba(0, 0, 0, 0.02)';
export const shadow_s2_right = '4px 0px 16px 4px rgba(0, 0, 0, 0.03), 4px 0px 8px 0px rgba(0, 0, 0, 0.02), 2px 0px 4px -4px rgba(0, 0, 0, 0.02)';
export const shadow_s2_up = '0px -4px 16px 4px rgba(0, 0, 0, 0.03), 0px -4px 8px 0px rgba(0, 0, 0, 0.02), 0px -2px 4px -4px rgba(0, 0, 0, 0.02)';
export const shadow_s3_down = '0px 6px 18px 6px rgba(0, 0, 0, 0.03), 0px 3px 6px -6px rgba(0, 0, 0, 0.05), 0px 4px 8px 0px rgba(0, 0, 0, 0.03)';
export const shadow_s3_down_pri = '0px 6px 18px 6px rgba(36, 91, 219, 0.03), 0px 3px 6px -6px rgba(36, 91, 219, 0.05), 0px 4px 8px 0px rgba(36, 91, 219, 0.03)';
export const shadow_s3_left = '-6px 0px 18px 6px rgba(0, 0, 0, 0.03), -3px 0px 6px -6px rgba(0, 0, 0, 0.05), -4px 0px 8px 0px rgba(0, 0, 0, 0.03)';
export const shadow_s3_right = '6px 0px 18px 6px rgba(0, 0, 0, 0.03), 3px 0px 6px -6px rgba(0, 0, 0, 0.05), 4px 0px 8px 0px rgba(0, 0, 0, 0.03)';
export const shadow_s3_up = '0px -6px 18px 6px rgba(0, 0, 0, 0.03), 0px -3px 6px -6px rgba(0, 0, 0, 0.05), 0px -4px 8px 0px rgba(0, 0, 0, 0.03)';
export const shadow_s4_down = '0px 8px 24px 8px rgba(0, 0, 0, 0.04), 0px 6px 12px 0px rgba(0, 0, 0, 0.04), 0px 4px 8px -8px rgba(0, 0, 0, 0.06)';
export const shadow_s4_down_pri = '0px 8px 24px 8px rgba(36, 91, 219, 0.04), 0px 6px 12px 0px rgba(36, 91, 219, 0.04), 0px 4px 8px -8px rgba(36, 91, 219, 0.06)';
export const shadow_s4_left = '-8px 0px 24px 8px rgba(0, 0, 0, 0.04), -6px 0px 12px 0px rgba(0, 0, 0, 0.04), -4px 0px 8px -8px rgba(0, 0, 0, 0.06)';
export const shadow_s4_right = '8px 0px 24px 8px rgba(0, 0, 0, 0.04), 6px 0px 12px 0px rgba(0, 0, 0, 0.04), 4px 0px 8px -8px rgba(0, 0, 0, 0.06)';
export const shadow_s4_up = '0px -8px 24px 8px rgba(0, 0, 0, 0.04), 0px -6px 12px 0px rgba(0, 0, 0, 0.04), 0px -4px 8px -8px rgba(0, 0, 0, 0.06)';
export const shadow_s5_down = '0px 10px 36px 10px rgba(0, 0, 0, 0.04), 0px 8px 24px 0px rgba(0, 0, 0, 0.04), 0px 6px 12px -10px rgba(0, 0, 0, 0.06)';
export const shadow_s5_down_pri = '0px 10px 36px 10px rgba(36, 91, 219, 0.04), 0px 8px 24px 0px rgba(36, 91, 219, 0.04), 0px 6px 12px -10px rgba(36, 91, 219, 0.06)';
export const shadow_s5_left = '-10px 0px 36px 10px rgba(0, 0, 0, 0.04), -8px 0px 24px 0px rgba(0, 0, 0, 0.04), -6px 0px 12px -10px rgba(0, 0, 0, 0.06)';
export const shadow_s5_right = '10px 0px 36px 10px rgba(0, 0, 0, 0.04), 8px 0px 24px 0px rgba(0, 0, 0, 0.04), 6px 0px 12px -10px rgba(0, 0, 0, 0.06)';
export const shadow_s5_up = '0px -10px 36px 10px rgba(0, 0, 0, 0.04), 0px -8px 24px 0px rgba(0, 0, 0, 0.04), 0px -6px 12px -10px rgba(0, 0, 0, 0.06)';
export const shadow_sm_down = '0px 4px 8px 0px rgba(0, 0, 0, 0.32)';
export const shadow_sm_left = '-4px 0px 8px 0px rgba(0, 0, 0, 0.32)';
export const shadow_sm_right = '4px 0px 8px 0px rgba(0, 0, 0, 0.32)';
export const shadow_sm_up = '0px -4px 8px 0px rgba(0, 0, 0, 0.32)';
export const static_black = 'rgb(0, 0, 0)';
export const static_black_10 = 'rgba(0, 0, 0, 0.1)';
export const static_black_15 = 'rgba(0, 0, 0, 0.15)';
export const static_black_20 = 'rgba(0, 0, 0, 0.2)';
export const static_black_30 = 'rgba(0, 0, 0, 0.3)';
export const static_black_40 = 'rgba(0, 0, 0, 0.4)';
export const static_black_5 = 'rgba(0, 0, 0, 0.05)';
export const static_black_50 = 'rgba(0, 0, 0, 0.5)';
export const static_black_60 = 'rgba(0, 0, 0, 0.6)';
export const static_black_70 = 'rgba(0, 0, 0, 0.7)';
export const static_black_80 = 'rgba(0, 0, 0, 0.8)';
export const static_black_90 = 'rgba(0, 0, 0, 0.9)';
export const static_white = 'rgb(255, 255, 255)';
export const static_white_10 = 'rgba(255, 255, 255, 0.1)';
export const static_white_15 = 'rgba(255, 255, 255, 0.15)';
export const static_white_20 = 'rgba(255, 255, 255, 0.2)';
export const static_white_30 = 'rgba(255, 255, 255, 0.3)';
export const static_white_40 = 'rgba(255, 255, 255, 0.4)';
export const static_white_5 = 'rgba(255, 255, 255, 0.05)';
export const static_white_50 = 'rgba(255, 255, 255, 0.5)';
export const static_white_60 = 'rgba(255, 255, 255, 0.6)';
export const static_white_70 = 'rgba(255, 255, 255, 0.7)';
export const static_white_80 = 'rgba(255, 255, 255, 0.8)';
export const static_white_90 = 'rgba(255, 255, 255, 0.9)';
export const static_white_hover = 'rgba(255, 255, 255, 0.1)';
export const static_white_pressed = 'rgba(255, 255, 255, 0.2)';
export const text_caption = 'rgb(166, 166, 166)';
export const text_disabled = 'rgb(95, 95, 95)';
export const text_link_disabled = 'rgb(95, 95, 95)';
export const text_link_hover = 'rgb(51, 112, 235)';
export const text_link_loading = 'rgb(38, 85, 182)';
export const text_link_normal = 'rgb(76, 136, 255)';
export const text_link_pressed = 'rgb(117, 164, 255)';
export const text_placeholder = 'rgb(117, 117, 117)';
export const text_title = 'rgb(235, 235, 235)';
