// @universe-design/styles/esm/preset-dark.atom-raw.js
// autogen by ../scripts/gen-ts-from-json.ts
export const AI_cursor_raw = '94, 104, 232';
export const AI_icon_in_button_raw = '123, 131, 247';
export const B100_raw = '23, 49, 102';
export const B200_raw = '25, 66, 148';
export const B300_raw = '38, 85, 182';
export const B350_raw = '39, 95, 206';
export const B400_raw = '51, 112, 235';
export const B50_raw = '21, 35, 64';
export const B500_raw = '76, 136, 255';
export const B600_raw = '117, 164, 255';
export const B700_raw = '143, 180, 255';
export const B800_raw = '189, 210, 255';
export const B900_raw = '224, 233, 255';
export const C100_raw = '89, 28, 63';
export const C200_raw = '120, 43, 87';
export const C300_raw = '148, 56, 108';
export const C350_raw = '171, 65, 125';
export const C400_raw = '194, 74, 142';
export const C50_raw = '58, 24, 43';
export const C500_raw = '219, 94, 164';
export const C600_raw = '237, 119, 186';
export const C700_raw = '252, 148, 207';
export const C800_raw = '255, 194, 229';
export const C900_raw = '255, 224, 242';
export const G100_raw = '23, 59, 18';
export const G200_raw = '33, 81, 26';
export const G300_raw = '41, 102, 33';
export const G350_raw = '47, 117, 38';
export const G400_raw = '53, 135, 42';
export const G50_raw = '14, 43, 10';
export const G500_raw = '65, 158, 52';
export const G600_raw = '81, 186, 67';
export const G700_raw = '105, 204, 92';
export const G800_raw = '153, 228, 144';
export const G900_raw = '203, 245, 198';
export const I100_raw = '42, 45, 105';
export const I200_raw = '55, 61, 144';
export const I300_raw = '71, 79, 184';
export const I350_raw = '81, 90, 214';
export const I400_raw = '94, 104, 232';
export const I50_raw = '30, 32, 74';
export const I500_raw = '123, 131, 247';
export const I600_raw = '148, 153, 247';
export const I700_raw = '170, 175, 248';
export const I800_raw = '205, 208, 249';
export const I900_raw = '231, 233, 254';
export const L100_raw = '48, 56, 4';
export const L200_raw = '64, 76, 6';
export const L300_raw = '83, 97, 10';
export const L350_raw = '95, 112, 10';
export const L400_raw = '107, 127, 5';
export const L50_raw = '33, 39, 2';
export const L500_raw = '130, 153, 10';
export const L600_raw = '147, 175, 4';
export const L700_raw = '166, 195, 19';
export const L800_raw = '194, 225, 45';
export const L900_raw = '227, 243, 145';
export const N00_raw = '10, 10, 10';
export const N100_raw = '41, 41, 41';
export const N1000_raw = '255, 255, 255';
export const N200_raw = '55, 55, 55';
export const N300_raw = '67, 67, 67';
export const N350_raw = '80, 80, 80';
export const N400_raw = '95, 95, 95';
export const N50_raw = '26, 26, 26';
export const N500_raw = '117, 117, 117';
export const N600_raw = '166, 166, 166';
export const N650_raw = '207, 207, 207';
export const N700_raw = '224, 224, 224';
export const N800_raw = '232, 232, 232';
export const N900_raw = '235, 235, 235';
export const N950_raw = '248, 248, 248';
export const O100_raw = '74, 43, 16';
export const O200_raw = '104, 58, 18';
export const O300_raw = '138, 74, 25';
export const O350_raw = '161, 83, 23';
export const O400_raw = '184, 94, 26';
export const O50_raw = '51, 33, 11';
export const O500_raw = '219, 112, 24';
export const O600_raw = '243, 135, 27';
export const O700_raw = '248, 158, 68';
export const O800_raw = '254, 200, 139';
export const O900_raw = '254, 231, 205';
export const P100_raw = '63, 32, 115';
export const P200_raw = '85, 41, 163';
export const P300_raw = '108, 57, 198';
export const P350_raw = '124, 74, 212';
export const P400_raw = '140, 85, 236';
export const P50_raw = '43, 25, 74';
export const P500_raw = '165, 117, 250';
export const P600_raw = '184, 143, 254';
export const P700_raw = '197, 163, 255';
export const P800_raw = '219, 200, 253';
export const P900_raw = '239, 229, 255';
export const R100_raw = '89, 31, 29';
export const R200_raw = '123, 37, 36';
export const R300_raw = '160, 51, 49';
export const R350_raw = '179, 58, 55';
export const R400_raw = '209, 70, 66';
export const R50_raw = '61, 26, 25';
export const R500_raw = '240, 91, 86';
export const R600_raw = '246, 130, 126';
export const R700_raw = '248, 158, 155';
export const R800_raw = '253, 198, 196';
export const R900_raw = '254, 227, 226';
export const S100_raw = '56, 50, 12';
export const S200_raw = '87, 77, 1';
export const S300_raw = '122, 106, 1';
export const S350_raw = '156, 135, 2';
export const S400_raw = '201, 178, 24';
export const S50_raw = '41, 37, 10';
export const S500_raw = '229, 205, 23';
export const S600_raw = '245, 223, 54';
export const S700_raw = '247, 230, 99';
export const S800_raw = '250, 237, 122';
export const S900_raw = '255, 245, 158';
export const T100_raw = '23, 59, 54';
export const T200_raw = '29, 78, 71';
export const T300_raw = '37, 102, 94';
export const T350_raw = '34, 115, 105';
export const T400_raw = '25, 133, 120';
export const T50_raw = '19, 41, 38';
export const T500_raw = '31, 161, 143';
export const T600_raw = '26, 183, 161';
export const T700_raw = '23, 207, 181';
export const T800_raw = '101, 231, 213';
export const T900_raw = '183, 247, 239';
export const V100_raw = '84, 24, 84';
export const V200_raw = '113, 40, 113';
export const V300_raw = '139, 55, 139';
export const V350_raw = '164, 61, 164';
export const V400_raw = '181, 74, 181';
export const V50_raw = '59, 21, 59';
export const V500_raw = '214, 97, 214';
export const V600_raw = '225, 127, 225';
export const V700_raw = '233, 155, 233';
export const V800_raw = '244, 195, 244';
export const V900_raw = '252, 223, 252';
export const Vnext_bg_content_header_raw = '34, 34, 34';
export const Vnext_bg_sub_navigation_raw = '30, 30, 30';
export const Vnext_bg_thirdly_navigation_raw = '26, 26, 26';
export const W100_raw = '16, 54, 71';
export const W200_raw = '22, 67, 89';
export const W300_raw = '19, 90, 120';
export const W350_raw = '20, 108, 145';
export const W400_raw = '26, 127, 171';
export const W50_raw = '21, 40, 48';
export const W500_raw = '16, 153, 204';
export const W600_raw = '37, 178, 229';
export const W700_raw = '81, 195, 240';
export const W800_raw = '137, 223, 254';
export const W900_raw = '199, 240, 255';
export const Y100_raw = '71, 52, 9';
export const Y200_raw = '99, 71, 15';
export const Y300_raw = '138, 100, 25';
export const Y350_raw = '173, 125, 21';
export const Y400_raw = '212, 155, 11';
export const Y50_raw = '48, 37, 10';
export const Y500_raw = '240, 182, 34';
export const Y600_raw = '251, 203, 70';
export const Y700_raw = '252, 212, 86';
export const Y800_raw = '255, 222, 117';
export const Y900_raw = '255, 234, 163';
export const bg_base_raw = '10, 10, 10';
export const bg_body_overlay_raw = '41, 41, 41';
export const bg_body_raw = '26, 26, 26';
export const bg_content_base_raw = '18, 18, 18';
export const bg_filler_raw = '55, 55, 55';
export const bg_float_base_raw = '26, 26, 26';
export const bg_float_overlay_raw = '55, 55, 55';
export const bg_float_raw = '41, 41, 41';
export const bg_mask_raw = '0, 0, 0';
export const bg_pricolor_raw = '51, 112, 235';
export const bg_sub_navigation_raw = '38, 38, 38';
export const bg_tips_raw = '80, 80, 80';
export const colorful_blue_raw = '76, 136, 255';
export const colorful_carmine_raw = '194, 74, 142';
export const colorful_green_raw = '65, 158, 52';
export const colorful_indigo_raw = '94, 104, 232';
export const colorful_lime_raw = '147, 175, 4';
export const colorful_neutral_raw = '117, 117, 117';
export const colorful_orange_raw = '243, 135, 27';
export const colorful_purple_raw = '140, 85, 236';
export const colorful_red_raw = '240, 91, 86';
export const colorful_sunflower_raw = '229, 205, 23';
export const colorful_turquoise_raw = '31, 161, 143';
export const colorful_violet_raw = '181, 74, 181';
export const colorful_wathet_raw = '16, 153, 204';
export const colorful_yellow_raw = '240, 182, 34';
export const fill_disabled_raw = '95, 95, 95';
export const function_danger_100_raw = '89, 31, 29';
export const function_danger_200_raw = '123, 37, 36';
export const function_danger_300_raw = '160, 51, 49';
export const function_danger_350_raw = '179, 58, 55';
export const function_danger_400_raw = '209, 70, 66';
export const function_danger_50_raw = '61, 26, 25';
export const function_danger_500_raw = '240, 91, 86';
export const function_danger_600_raw = '246, 130, 126';
export const function_danger_700_raw = '248, 158, 155';
export const function_danger_800_raw = '253, 198, 196';
export const function_danger_900_raw = '254, 227, 226';
export const function_danger_content_default_raw = '240, 91, 86';
export const function_danger_content_hover_raw = '209, 70, 66';
export const function_danger_content_loading_raw = '160, 51, 49';
export const function_danger_content_pressed_raw = '246, 130, 126';
export const function_danger_fill_default_raw = '209, 70, 66';
export const function_danger_fill_hover_raw = '179, 58, 55';
export const function_danger_fill_loading_raw = '160, 51, 49';
export const function_danger_fill_pressed_raw = '240, 91, 86';
export const function_danger_fill_solid_01_raw = '61, 26, 25';
export const function_danger_fill_solid_02_raw = '89, 31, 29';
export const function_danger_fill_solid_03_raw = '123, 37, 36';
export const function_danger_on_danger_fill_raw = '255, 255, 255';
export const function_info_100_raw = '23, 49, 102';
export const function_info_200_raw = '25, 66, 148';
export const function_info_300_raw = '38, 85, 182';
export const function_info_350_raw = '39, 95, 206';
export const function_info_400_raw = '51, 112, 235';
export const function_info_50_raw = '21, 35, 64';
export const function_info_500_raw = '76, 136, 255';
export const function_info_600_raw = '117, 164, 255';
export const function_info_700_raw = '143, 180, 255';
export const function_info_800_raw = '189, 210, 255';
export const function_info_900_raw = '224, 233, 255';
export const function_info_content_default_raw = '76, 136, 255';
export const function_info_content_hover_raw = '51, 112, 235';
export const function_info_content_loading_raw = '38, 85, 182';
export const function_info_content_pressed_raw = '117, 164, 255';
export const function_info_fill_default_raw = '51, 112, 235';
export const function_info_fill_hover_raw = '39, 95, 206';
export const function_info_fill_loading_raw = '38, 85, 182';
export const function_info_fill_pressed_raw = '76, 136, 255';
export const function_info_fill_solid_01_raw = '21, 35, 64';
export const function_info_fill_solid_02_raw = '23, 49, 102';
export const function_info_fill_solid_03_raw = '25, 66, 148';
export const function_info_on_info_fill_raw = '255, 255, 255';
export const function_success_100_raw = '23, 59, 18';
export const function_success_200_raw = '33, 81, 26';
export const function_success_300_raw = '41, 102, 33';
export const function_success_350_raw = '47, 117, 38';
export const function_success_400_raw = '53, 135, 42';
export const function_success_50_raw = '14, 43, 10';
export const function_success_500_raw = '65, 158, 52';
export const function_success_600_raw = '81, 186, 67';
export const function_success_700_raw = '105, 204, 92';
export const function_success_800_raw = '153, 228, 144';
export const function_success_900_raw = '203, 245, 198';
export const function_success_content_default_raw = '81, 186, 67';
export const function_success_content_hover_raw = '65, 158, 52';
export const function_success_content_loading_raw = '41, 102, 33';
export const function_success_content_pressed_raw = '105, 204, 92';
export const function_success_fill_default_raw = '65, 158, 52';
export const function_success_fill_hover_raw = '53, 135, 42';
export const function_success_fill_loading_raw = '41, 102, 33';
export const function_success_fill_pressed_raw = '81, 186, 67';
export const function_success_fill_solid_01_raw = '14, 43, 10';
export const function_success_fill_solid_02_raw = '23, 59, 18';
export const function_success_fill_solid_03_raw = '33, 81, 26';
export const function_success_on_success_fill_raw = '255, 255, 255';
export const function_warning_100_raw = '74, 43, 16';
export const function_warning_200_raw = '104, 58, 18';
export const function_warning_300_raw = '138, 74, 25';
export const function_warning_350_raw = '161, 83, 23';
export const function_warning_400_raw = '184, 94, 26';
export const function_warning_50_raw = '51, 33, 11';
export const function_warning_500_raw = '219, 112, 24';
export const function_warning_600_raw = '243, 135, 27';
export const function_warning_700_raw = '248, 158, 68';
export const function_warning_800_raw = '254, 200, 139';
export const function_warning_900_raw = '254, 231, 205';
export const function_warning_content_default_raw = '243, 135, 27';
export const function_warning_content_hover_raw = '219, 112, 24';
export const function_warning_content_loading_raw = '138, 74, 25';
export const function_warning_content_pressed_raw = '248, 158, 68';
export const function_warning_fill_default_raw = '243, 135, 27';
export const function_warning_fill_hover_raw = '219, 112, 24';
export const function_warning_fill_loading_raw = '138, 74, 25';
export const function_warning_fill_pressed_raw = '248, 158, 68';
export const function_warning_fill_solid_01_raw = '51, 33, 11';
export const function_warning_fill_solid_02_raw = '74, 43, 16';
export const function_warning_fill_solid_03_raw = '104, 58, 18';
export const function_warning_on_warning_fill_raw = '255, 255, 255';
export const icon_disabled_raw = '95, 95, 95';
export const icon_n1_raw = '232, 232, 232';
export const icon_n2_raw = '166, 166, 166';
export const icon_n3_raw = '117, 117, 117';
export const illustration_blue_a_raw = '224, 224, 224';
export const illustration_blue_b_raw = '255, 255, 255';
export const illustration_blue_c_raw = '95, 95, 95';
export const illustration_blue_d_raw = '38, 85, 182';
export const illustration_blue_e_raw = '76, 136, 255';
export const illustration_neutral_a_raw = '235, 235, 235';
export const illustration_neutral_b_raw = '55, 55, 55';
export const illustration_neutral_d_raw = '255, 255, 255';
export const illustration_orange_raw = '243, 135, 27';
export const illustration_red_raw = '240, 91, 86';
export const illustration_static_black_raw = '0, 0, 0';
export const illustration_static_white_raw = '255, 255, 255';
export const illustration_turquoise_raw = '23, 207, 181';
export const illustration_wathet_raw = '81, 195, 240';
export const illustration_yellow_raw = '240, 182, 34';
export const inverse_primary_content_default_raw = '117, 164, 255';
export const inverse_primary_content_hover_raw = '143, 180, 255';
export const inverse_primary_content_pressed_raw = '76, 136, 255';
export const line_border_card_raw = '235, 235, 235';
export const line_border_component_raw = '80, 80, 80';
export const primary_content_default_raw = '76, 136, 255';
export const primary_content_hover_raw = '51, 112, 235';
export const primary_content_loading_raw = '38, 85, 182';
export const primary_content_pressed_raw = '117, 164, 255';
export const primary_fill_default_raw = '51, 112, 235';
export const primary_fill_hover_raw = '39, 95, 206';
export const primary_fill_loading_raw = '38, 85, 182';
export const primary_fill_pressed_raw = '76, 136, 255';
export const primary_fill_solid_01_raw = '21, 35, 64';
export const primary_fill_solid_02_raw = '23, 49, 102';
export const primary_fill_solid_03_raw = '25, 66, 148';
export const primary_on_primary_fill_raw = '255, 255, 255';
export const primary_pri_100_raw = '23, 49, 102';
export const primary_pri_200_raw = '25, 66, 148';
export const primary_pri_300_raw = '38, 85, 182';
export const primary_pri_350_raw = '39, 95, 206';
export const primary_pri_400_raw = '51, 112, 235';
export const primary_pri_50_raw = '21, 35, 64';
export const primary_pri_500_raw = '76, 136, 255';
export const primary_pri_600_raw = '117, 164, 255';
export const primary_pri_700_raw = '143, 180, 255';
export const primary_pri_800_raw = '189, 210, 255';
export const primary_pri_900_raw = '224, 233, 255';
export const shadow_default_raw = '0, 0, 0';
export const shadow_pri_raw = '36, 91, 219';
export const static_black_raw = '0, 0, 0';
export const static_white_hover_raw = '255, 255, 255';
export const static_white_pressed_raw = '255, 255, 255';
export const static_white_raw = '255, 255, 255';
export const text_caption_raw = '166, 166, 166';
export const text_disabled_raw = '95, 95, 95';
export const text_link_disabled_raw = '95, 95, 95';
export const text_link_hover_raw = '51, 112, 235';
export const text_link_loading_raw = '38, 85, 182';
export const text_link_normal_raw = '76, 136, 255';
export const text_link_pressed_raw = '117, 164, 255';
export const text_placeholder_raw = '117, 117, 117';
export const text_title_raw = '235, 235, 235';
