// @universe-design/react/lib/breadcrumb/BreadcrumbItem.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BreadcrumbItem = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _hooks = require("../config-provider/hooks");
var _text = require("../text");
require("./style");
const BreadcrumbItem = exports.BreadcrumbItem = /*#__PURE__*/_react.default.forwardRef((_ref, ref) => {
  let {
    className,
    children,
    interactive = true,
    maxWidth,
    style,
    ...otherProps
  } = _ref;
  const cssScope = (0, _hooks.useCssScope)('breadcrumb-item');
  const mergedStyle = {
    ...style,
    maxWidth
  };
  return /*#__PURE__*/_react.default.createElement(_text.Text, (0, _extends2.default)({
    as: "div",
    className: (0, _clsx.default)(cssScope(), className, interactive && cssScope('-interactive')),
    ref: ref,
    style: mergedStyle,
    lineClamp: 1
  }, otherProps), children);
});
if (process.env.NODE_ENV !== "production") {
  BreadcrumbItem.displayName = 'BreadcrumbItem';
}
