// @universe-design/react/lib/breadcrumb/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Breadcrumb = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _MoreOutlined = _interopRequireDefault(require("@universe-design/icons-react/MoreOutlined"));
var _RightBoldOutlined = _interopRequireDefault(require("@universe-design/icons-react/RightBoldOutlined"));
var _hooks = require("../config-provider/hooks");
var _BreadcrumbItem = require("./BreadcrumbItem");
var _popover = require("../popover");
var _overflow = require("../overflow");
var _util = require("../util");
require("./style");
const Breadcrumb = exports.Breadcrumb = Object.assign(/*#__PURE__*/_react.default.forwardRef((props, ref) => {
  const {
    className,
    children,
    maxCount = Infinity,
    separator = /*#__PURE__*/_react.default.createElement(_RightBoldOutlined.default, null),
    ...restProps
  } = props;
  const cssScope = (0, _hooks.useCssScope)('breadcrumb');
  const childNodes = (0, _util.toFlatArray)(children);
  const lastChildNode = childNodes.pop();
  const separatorNode = /*#__PURE__*/_react.default.createElement("span", {
    className: cssScope('-separator')
  }, separator);
  const [visible, setVisible] = (0, _react.useState)(false);
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
    className: (0, _clsx.default)(cssScope(), className),
    ref: ref
  }, restProps), /*#__PURE__*/_react.default.createElement(_overflow.Overflow, {
    restPosition: 1,
    maxCount: typeof maxCount === 'number' ? maxCount - 1 : 'responsive',
    keys: Array(childNodes.length).fill(0).map((_, i) => i),
    itemRenderer: key => /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, childNodes[key], separatorNode),
    restRenderer: omittedKeys => /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_popover.Popover, {
      mode: "pure",
      visible: visible,
      content: /*#__PURE__*/_react.default.createElement("div", {
        onClick: () => setVisible(false),
        className: cssScope('__popover__content')
      }, omittedKeys.map((o, i) => /*#__PURE__*/_react.default.createElement(_react.default.Fragment, {
        key: o
      }, childNodes[o], i !== omittedKeys.length - 1 && separatorNode))),
      placement: "bottomLeft",
      hideArrow: true,
      onVisibleChange: setVisible
    }, /*#__PURE__*/_react.default.createElement(_BreadcrumbItem.BreadcrumbItem, {
      className: cssScope('-ellipses')
    }, /*#__PURE__*/_react.default.createElement(_MoreOutlined.default, null))), separatorNode),
    suffixRenderer: () => lastChildNode ? /*#__PURE__*/(0, _react.cloneElement)(lastChildNode, {
      interactive: false
    }) : null
  }));
}), {
  Item: _BreadcrumbItem.BreadcrumbItem
});
if (process.env.NODE_ENV !== "production") {
  Breadcrumb.displayName = 'Breadcrumb';
}
