// @universe-design/react/lib/input-number/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InputNumber = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _AddOutlined = _interopRequireDefault(require("@universe-design/icons-react/AddOutlined"));
var _UpBoldOutlined = _interopRequireDefault(require("@universe-design/icons-react/UpBoldOutlined"));
var _DownBoldOutlined = _interopRequireDefault(require("@universe-design/icons-react/DownBoldOutlined"));
var _ReduceOutlined = _interopRequireDefault(require("@universe-design/icons-react/ReduceOutlined"));
var _reactUtilities = require("@universe-design/react-utilities");
var _util = require("../util");
var _hooks = require("../config-provider/hooks");
var _noop = _interopRequireDefault(require("lodash/noop"));
var _Input = _interopRequireDefault(require("../shared/components/Input"));
var _useSyncState = _interopRequireDefault(require("../hooks/useSyncState"));
var _reactInput = require("@universe-design/react-input");
var _utils = require("./utils");
require("./style");
// nolint-next-line: long_function
const InputNumber = exports.InputNumber = /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
  const defaultSize = (0, _hooks.useComponentSize)();
  const inputGroupSize = (0, _hooks.useInputGroupSize)();
  const {
    min = -_utils.MAX_SAFE_INTEGER,
    max = _utils.MAX_SAFE_INTEGER,
    size = inputGroupSize || defaultSize,
    step = 1,
    mode = 'default',
    error,
    readOnly,
    disabled,
    autoFocus,
    precision,
    placeholder,
    onlyNumber = false,
    value: valueInProps,
    defaultValue,
    className,
    style,
    addonAfter,
    addonBefore,
    formatter: formatterProp,
    parser = formatterProp ? input => input.replace(/[^\w.-]+/g, '') : undefined,
    onChange,
    onPressEnter,
    onBlur,
    onFocus,
    onMouseEnter,
    onMouseLeave,
    onClick,
    onStep,
    controls = true
  } = props;
  const formatter = formatterProp !== null && formatterProp !== void 0 ? formatterProp : input => input;
  const limitWithMinMax = value => {
    let val = value;
    if ((max || max === 0) && value > max) {
      val = max;
    }
    if ((min || min === 0) && val < min) {
      val = min;
    }
    return val;
  };

  /**
   * toFixed 方法不能保证准确地做四舍五入，
   * 所以需要四舍五入地做保留小数位必须使用此方法
   */
  const processPrecision = (n, currentPrecision) => {
    const splitNumber = `${n}`.split('.');
    const integerStr = splitNumber[0] || '0';
    const decimalStr = splitNumber[1] || '0';
    const advancedNum = decimalStr[currentPrecision] ? Number(decimalStr[currentPrecision]) : undefined;
    if (n > 0 && advancedNum && advancedNum >= 5) {
      return n + Number(`0.${'0'.repeat(currentPrecision)}${10 - advancedNum}`);
    }
    if (n < 0 && advancedNum) {
      return Number(`${integerStr}.${decimalStr.slice(0, currentPrecision)}`);
    }
    return n;
  };
  const calculateVal = (rawValue, type) => {
    if (rawValue === undefined || Number.isNaN(rawValue)) {
      if (type === _utils.ChangeType.submit) {
        return undefined;
      } else {
        rawValue = _utils.DEFAULT_VALUE;
      }
    }
    const rawPrecision = (0, _utils.getPrecision)(rawValue);
    const minPrecision = (0, _utils.getPrecision)(min);
    const maxPrecision = (0, _utils.getPrecision)(max);
    const stepPrecision = (0, _utils.getPrecision)(step);
    let steppedValue = rawValue;
    switch (type) {
      case _utils.ChangeType.addition:
        steppedValue = rawValue + step;
        break;
      case _utils.ChangeType.subtraction:
        steppedValue = rawValue - step;
        break;
      default:
        break;
    }
    const result = limitWithMinMax(steppedValue);
    /**
     *  计算 precision
     *  受控的 precision 优先级最高
     *  其次看当前值和 step 的 precision, 取最大值
     *  */
    const currentPrecision = (0, _util.isNonNullable)(precision) ? precision : Math.max(...[rawPrecision, stepPrecision, result === min ? minPrecision : null, result === max ? maxPrecision : null].filter(_util.isNonNullable));
    // 这里最后的 toFixed 只是为了补 0
    return processPrecision(result, currentPrecision).toFixed(currentPrecision);
  };
  const inputRef = (0, _react.useRef)(null);
  const timeout = (0, _react.useRef)();
  const inputingRef = (0, _react.useRef)(false);
  const compositionRef = (0, _react.useRef)({
    isComposing: false,
    nextValue: null
  });
  const [focused, setFocused] = (0, _react.useState)(Boolean(autoFocus));
  const [handleVisible, setHandleVisible] = (0, _react.useState)(false);
  const [recordCursor, resetCursor] = (0, _reactInput.useCursor)(inputRef, focused);
  const [isError, setError] = (0, _reactUtilities.useControlledState)(false, Boolean(error), {
    isControlledOverride: 'error' in props
  });

  /**
   * @param v 原始值
   * 将原始值转换成 number 类型的 value
   */
  function rawToValidValue(v) {
    const toNum = (0, _utils.getValidValueNumberFromAny)(v);
    if (Number.isNaN(toNum)) {
      return undefined;
    }
    const valueFormatted = calculateVal(toNum, _utils.ChangeType.submit);
    return valueFormatted === undefined ? undefined : Number(valueFormatted);
  }
  const [value, setValue] = (0, _reactUtilities.useControlledState)(rawToValidValue(defaultValue), rawToValidValue(valueInProps), {
    isControlledOverride: 'value' in props,
    onChange
  });

  /**
   * 将 value 转换成 viewValue (输入框显示的值)
   * @param v
   */
  function valueToViewValue(v) {
    if (v === undefined) {
      return '';
    }
    const valueStr = calculateVal(v, _utils.ChangeType.submit);
    return formatter(valueStr !== null && valueStr !== void 0 ? valueStr : '');
  }

  // 输入框显示的值，可能是非法的值或者被 formatter 格式化后的值
  const [viewValue, _setViewValue] = (0, _useSyncState.default)(valueToViewValue(value), {
    isEqual(v, preV) {
      const parsed = parser ? parser(v) : v;
      const preParsed = parser ? parser(preV) : preV;
      return rawToValidValue(preParsed) === rawToValidValue(parsed);
    }
  });

  /**
   * composition 参考文档：https://bytedance.larkoffice.com/wiki/XU9ewGUVViCeISkLOVpcvpbNnKd
   * 出问题的老版本微软输入法设置：https://jingyan.baidu.com/article/c1a3101eac4ca19f646deb07.html
   * 会打断组合输入状态的情况：
   * 1. 失去焦点，可能会触发两次 start+end；
   * 2. 切换输入法，老式微软输入法时，中文状态下输入 d，ctrl+shift 会导致 input 展示 $dd（format的值【$d】 + 输入法的值【d】），拼接展示，其他输入法正常
   * 3. 空格/enter 键，相当于选取
   * 4. 直接输入数字，本次重点修复 case，format 赋值不再打断组合输入状态，不会造成输入数字重复的问题
   *  */
  const setViewValue = v => {
    const input = inputRef.current;
    if (!input) {
      return;
    }
    // format 之后的值
    const nextValue = v === undefined ? '' : v.toString();
    // 组合输入状态时，input 的值与 format 之后的值不一致
    if (compositionRef.current.isComposing && input.value !== nextValue) {
      compositionRef.current.nextValue = nextValue;
      // 将 input 的值改为记录的值，防止打断组合输入状态，end 时再把 format 之后的值赋值
      _setViewValue(input.value);
      return;
    }
    /**
     * case:
     * 1. 非组合状态时
     * 2. 组合状态时，input 的值与 format 之后的值一致（比如: $12 输入 3， input.value 和 format 之后都是 $123）
     */
    compositionRef.current.nextValue = null;
    // 一致时，可以直接把 format 之后的赋值，跳过 onCompositionEnd 的赋值
    _setViewValue(nextValue);
  };

  // 将 viewValue 同步至 value，可能会触发 onChange
  /**
   * @param _viewValue 新提交的值
   * @param type 操作类型
   */
  function commitChange(_viewValue, type) {
    const parsedValue = parser ? parser(_viewValue.toString()) : _viewValue.toString();
    const flush = !inputingRef.current; // 是否要提交及处理 max/min/precision
    let newIsError = isError;
    if ((0, _utils.internalIsNumber)(parsedValue) || parsedValue === '') {
      newIsError = false;
    } else if (!onlyNumber) {
      // 如果 parse 出来的值不合法，标记为 error
      newIsError = true;
    }
    const toNum = (0, _utils.getValidValueNumberFromAny)(parsedValue);
    const valueWithPer = calculateVal(toNum, type);
    let newViewValue, newValue;
    if (valueWithPer === undefined) {
      // 判断formatter的参数 -> parsedValue / ""
      newViewValue = inputingRef.current ? formatter(parsedValue) : formatter('');
      newValue = undefined;
    } else {
      newViewValue = formatter(!flush ? parsedValue : valueWithPer);
      newValue = Number(valueWithPer);
    }
    if (flush) {
      newIsError = false;
    }
    if (flush || formatterProp) {
      setViewValue(newViewValue);
    } else {
      setViewValue(_viewValue);
    }
    setValue(newValue);
    if (newIsError !== isError) {
      setError(newIsError);
    }
    return newValue;
  }
  (0, _react.useImperativeHandle)(ref, () => inputRef.current);
  const stop = (0, _react.useCallback)(() => {
    timeout.current && clearTimeout(timeout.current);
  }, []);
  (0, _react.useEffect)(() => {
    if (focused && document.activeElement !== inputRef.current) {
      var _inputRef$current;
      (_inputRef$current = inputRef.current) === null || _inputRef$current === void 0 ? void 0 : _inputRef$current.focus();
    }
  });
  (0, _react.useEffect)(() => () => stop(), [stop]);
  (0, _react.useEffect)(() => {
    resetCursor();
  }, [resetCursor]);
  const disableUp = max || max === 0 ? value === max : false;
  const disableDown = min || min === 0 ? value === min : false;

  // css start
  const cssScope = (0, _hooks.useCssScope)('input-number');
  const usedSize = (0, _util.getDefaultSizeAbbr)(size);
  const hasAddon = addonAfter || addonBefore;
  const cls = (0, _clsx.default)(cssScope(), cssScope(`--mode-${mode}`), {
    [cssScope(`--size-${usedSize}`)]: usedSize,
    [cssScope('--focused')]: focused,
    [cssScope('--disabled')]: disabled,
    [cssScope('--readOnly')]: readOnly,
    [cssScope('--error')]: isError,
    [cssScope('--withoutContorls')]: !controls
  }, className);
  const inputCls = (0, _clsx.default)(cssScope('__input-wrap'));
  const addonWarp = hasAddon ? (0, _clsx.default)(cssScope('__addon'), {
    [cssScope('__addon-before-warp')]: addonBefore,
    [cssScope('__addon-after-warp')]: addonAfter,
    [cssScope('--disabled')]: disabled
  }, className) : '';
  const handleCls = (0, _clsx.default)(cssScope('__handle'), {
    [cssScope('__handle--visible')]: handleVisible
  });
  const arrowUpCls = (0, _clsx.default)(cssScope('__handle-item'), {
    [cssScope('__handle-item--disabled')]: disableUp
  });
  const arrowDownCls = (0, _clsx.default)(cssScope('__handle-item'), {
    [cssScope('__handle-item--disabled')]: disableDown
  });
  const stepperUpCls = (0, _clsx.default)(cssScope('__stepper-btn'), {
    [cssScope('__stepper-btn--disabled')]: disableUp
  });
  const stepperDownCls = (0, _clsx.default)(cssScope('__stepper-btn'), {
    [cssScope('__stepper-btn--disabled')]: disableDown
  });
  // css end
  const doStep = (inputValue, type, recursive) => {
    stop();
    inputingRef.current = false;
    const newValue = commitChange(inputValue, type); // 这里可以认为 step 操作出来的结果是个 number
    const typeStr = type === _utils.ChangeType.addition ? 'up' : 'down';
    (0, _util.isFunction)(onStep) && onStep(newValue, {
      offset: step,
      type: typeStr
    });
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    timeout.current = setTimeout(() => doStep(valueToViewValue(newValue), type, true), recursive ? 200 : 600);
  };
  const handleWrapMouseEnter = e => {
    (0, _util.isFunction)(onMouseEnter) && onMouseEnter(e);
    setHandleVisible(!disabled && !readOnly);
  };
  const handleWrapMouseLeave = e => {
    (0, _util.isFunction)(onMouseLeave) && onMouseLeave(e);
    setHandleVisible(false);
  };
  const handleInputKeyUp = () => {
    stop();
  };
  const handleInputKeyDown = e => {
    if (!inputRef.current) {
      return;
    }
    const inputValue = inputRef.current.value;
    if (e.keyCode === _util.KeyCode.UP) {
      e.preventDefault();
      doStep(inputValue, _utils.ChangeType.addition);
    } else if (e.keyCode === _util.KeyCode.DOWN) {
      e.preventDefault();
      doStep(inputValue, _utils.ChangeType.subtraction);
    } else if (e.keyCode === _util.KeyCode.ENTER) {
      inputingRef.current = false;
      commitChange(inputValue, _utils.ChangeType.submit);
      (0, _util.isFunction)(onPressEnter) && onPressEnter(e);
    }
  };
  const handleInputFocus = e => {
    (0, _util.isFunction)(onFocus) && onFocus(e);
    setFocused(true);
  };
  const handleInputBlur = e => {
    // 当 inputNumber 放在 Menu 组件里头, 如果按了 Up/Down 键, 会 focus 到 Menu 的其他菜单项
    // 导致没有触发 keyUp 事件, 所以就没触发到 keyUp 里的 stop 方法, 导致 input 中的数字不停自增或自减
    stop();
    (0, _util.isFunction)(onBlur) && onBlur(e);
    setFocused(false);
    inputingRef.current = false;
    commitChange(e.target.value, _utils.ChangeType.submit);
  };
  const handleUpClick = e => {
    /* istanbul ignore else  */
    if (!inputRef.current || disableUp) {
      return;
    }
    e.preventDefault();
    e.stopPropagation();
    if (!focused) {
      setFocused(true);
    }
    doStep(inputRef.current.value, _utils.ChangeType.addition);
  };
  const handleDownClick = e => {
    if (!inputRef.current || disableDown) {
      return;
    }
    e.preventDefault();
    e.stopPropagation();
    if (!focused) {
      setFocused(true);
    }
    doStep(inputRef.current.value, _utils.ChangeType.subtraction);
  };
  const handleChange = e => {
    inputingRef.current = true;
    recordCursor();
    // 始终把用户的输入更新到 input 中
    commitChange(e.target.value, _utils.ChangeType.submit);
  };

  /**
   * 禁止滚轮事件
   * input[type=number] 时，
   * 有用户填完数字以后有滚动页面页面的操作，当时输入框还没有失焦，所以数字发生了改变
   * 不使用 onWheel 的原因是 https://github.com/facebook/react/pull/19654
   */
  _react.default.useEffect(() => {
    const stopWheel = e => e.preventDefault();
    const input = inputRef.current;
    if (!input) {
      return;
    }
    input.addEventListener('wheel', stopWheel, {
      passive: false
    });
    return () => input.removeEventListener('wheel', stopWheel);
  }, []);
  const otherProps = onlyNumber ? {
    type: 'number'
  } : {};
  const renderInput = () => /*#__PURE__*/_react.default.createElement("div", {
    className: inputCls
  }, /*#__PURE__*/_react.default.createElement(_Input.default, (0, _extends2.default)({
    ref: inputRef,
    placeholder: placeholder,
    value: viewValue,
    onChange: handleChange,
    onKeyUp: disabled ? _noop.default : handleInputKeyUp,
    onKeyDown: disabled || readOnly ? _noop.default : handleInputKeyDown,
    onFocus: disabled || readOnly ? _noop.default : handleInputFocus,
    onBlur: disabled || readOnly ? _noop.default : handleInputBlur,
    autoComplete: "off",
    readOnly: readOnly,
    autoFocus: autoFocus,
    disabled: disabled,
    onCompositionStart: () => {
      compositionRef.current.isComposing = true;
    },
    onCompositionEnd: () => {
      // 组合输入状态结束时，把 format 的值更新到 input 中
      if (compositionRef.current.isComposing && compositionRef.current.nextValue) {
        _setViewValue(compositionRef.current.nextValue);
      }
      compositionRef.current.isComposing = false;
      compositionRef.current.nextValue = null;
    }
  }, otherProps)));
  const renderBody = () => {
    if (mode === 'stepper') {
      return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, !readOnly && /*#__PURE__*/_react.default.createElement("div", {
        className: stepperDownCls,
        onMouseDown: disabled ? _noop.default : handleDownClick,
        onMouseUp: disabled ? _noop.default : stop,
        onMouseLeave: disabled ? _noop.default : stop
      }, /*#__PURE__*/_react.default.createElement(_ReduceOutlined.default, null)), renderInput(), !readOnly && /*#__PURE__*/_react.default.createElement("div", {
        className: stepperUpCls,
        onMouseDown: disabled ? _noop.default : handleUpClick,
        onMouseUp: disabled ? _noop.default : stop,
        onMouseLeave: disabled ? _noop.default : stop
      }, /*#__PURE__*/_react.default.createElement(_AddOutlined.default, null)));
    }
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, renderInput(), !readOnly && controls && /*#__PURE__*/_react.default.createElement("div", {
      className: handleCls
    }, /*#__PURE__*/_react.default.createElement("span", {
      className: arrowUpCls,
      onMouseDown: disabled ? _noop.default : handleUpClick,
      onMouseUp: disabled ? _noop.default : stop,
      onMouseLeave: disabled ? _noop.default : stop
    }, /*#__PURE__*/_react.default.createElement(_UpBoldOutlined.default, null)), /*#__PURE__*/_react.default.createElement("span", {
      className: arrowDownCls,
      onMouseDown: disabled ? _noop.default : handleDownClick,
      onMouseUp: disabled ? _noop.default : stop,
      onMouseLeave: disabled ? _noop.default : stop
    }, /*#__PURE__*/_react.default.createElement(_DownBoldOutlined.default, null))));
  };
  const renderAddonBefore = function () {
    let prefix = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : cssScope();
    if (addonBefore) {
      return /*#__PURE__*/_react.default.createElement("span", {
        className: `${prefix}__addon-before`
      }, addonBefore);
    }
    return null;
  };
  const renderAddonAfter = function () {
    let prefix = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : cssScope();
    if (addonAfter) {
      return /*#__PURE__*/_react.default.createElement("span", {
        className: `${prefix}__addon-after`
      }, addonAfter);
    }
    return null;
  };
  if (hasAddon) {
    return /*#__PURE__*/_react.default.createElement("div", {
      className: addonWarp,
      onMouseEnter: disabled ? _noop.default : handleWrapMouseEnter,
      onMouseLeave: disabled ? _noop.default : handleWrapMouseLeave,
      onClick: onClick
    }, renderAddonBefore(), /*#__PURE__*/_react.default.createElement("div", {
      className: cls,
      style: style
    }, renderBody()), renderAddonAfter());
  }
  return /*#__PURE__*/_react.default.createElement("div", {
    className: cls,
    style: style,
    onMouseEnter: disabled ? _noop.default : handleWrapMouseEnter,
    onMouseLeave: disabled ? _noop.default : handleWrapMouseLeave,
    onClick: onClick
  }, renderBody());
});
