// @universe-design/react/lib/input-number/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.internalIsNumber = exports.getValidValueNumberFromAny = exports.getPrecision = exports.MAX_SAFE_INTEGER = exports.DEFAULT_VALUE = exports.ChangeType = void 0;
var _util = require("../util");
// eslint-disable-next-line @typescript-eslint/no-magic-numbers
const MAX_SAFE_INTEGER = exports.MAX_SAFE_INTEGER = Number.MAX_SAFE_INTEGER || 2 ** 53 - 1;

/**
 * 当直接使用 up/down 时
 *  1. 如果设置了 min, 则将值切换到 min
 *  2. 如果没有设置 min, 则将值变为 1
 */
const DEFAULT_VALUE = exports.DEFAULT_VALUE = 0;
const internalIsNumber = val => val !== '' && !isNaN(Number(val));
exports.internalIsNumber = internalIsNumber;
const getNumbersReg = /[.+-]?(?!0\d)\d+(\.\d+)?(E[+-]?\d+)?/gi;
const getValidValueNumberFromAny = val => {
  if (internalIsNumber(val)) {
    return Number(val);
  }
  if ((0, _util.isString)(val)) {
    const tempRes = val.match(getNumbersReg);
    return tempRes ? Number(tempRes[0]) : NaN;
  }
  return NaN;
};
exports.getValidValueNumberFromAny = getValidValueNumberFromAny;
const getPrecision = value => {
  const valueString = String(value);
  if (valueString.indexOf('e-') >= 0) {
    return parseInt(valueString.slice(valueString.indexOf('e-') + 2), 10);
  }
  let result = 0;
  if (valueString.indexOf('.') >= 0) {
    result = valueString.length - valueString.indexOf('.') - 1;
  }
  return result;
};
exports.getPrecision = getPrecision;
let ChangeType = exports.ChangeType = /*#__PURE__*/function (ChangeType) {
  ChangeType[ChangeType["submit"] = 0] = "submit";
  ChangeType[ChangeType["addition"] = 1] = "addition";
  ChangeType[ChangeType["subtraction"] = 2] = "subtraction";
  return ChangeType;
}({});
