// @universe-design/react/src/breadcrumb/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0_1Basic from '../demo/0-1.basic';
import C0Basic from '../demo/0.basic';
import C1ItemOverflow from '../demo/1.item-overflow';
import C2MaxCount from '../demo/2.max-count';
import C3MaxWidth from '../demo/3.max-width';
if (!(C0_1Basic as Story).storyName) {
  (C0_1Basic as Story).storyName = '0.1. 自定义分隔符';
  (C0_1Basic as Story).parameters = {
    docs: {
      description: {
        story: '通过`separator`自定义分隔符。\n\n',
      },
    },
  };
}
if (!(C0Basic as Story).storyName) {
  (C0Basic as Story).storyName = '0. 基础用法';
  (C0Basic as Story).parameters = {
    docs: {
      description: {
        story: '简单的用法。\n\n',
      },
    },
  };
}
if (!(C1ItemOverflow as Story).storyName) {
  (C1ItemOverflow as Story).storyName = '1. 单个条目省略';
  (C1ItemOverflow as Story).parameters = {
    docs: {
      description: {
        story: '当单个条目的宽度大于`162px`时，此条目会省略显示。\n\n',
      },
    },
  };
}
if (!(C2MaxCount as Story).storyName) {
  (C2MaxCount as Story).storyName = '2. 面包屑显示省略';
  (C2MaxCount as Story).parameters = {
    docs: {
      description: {
        story:
          '通过`maxCount`来指定最多渲染的面包屑条目数量，超出的部分将显示为省略号。`maxCount` 为 `responsive` 时会根据容器宽度自动收起\n\n',
      },
    },
  };
}
if (!(C3MaxWidth as Story).storyName) {
  (C3MaxWidth as Story).storyName = '3. 最大宽度';
  (C3MaxWidth as Story).parameters = {
    docs: {
      description: {
        story: '通过`maxWidth`来指定面包屑的最大宽度，如果不设置，最大宽度为 162px。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Breadcrumb/官网示例',
  tags: ['autodocs'],
};

export { C0_1Basic, C0Basic, C1ItemOverflow, C2MaxCount, C3MaxWidth };
