// @universe-design/react/lib/drawer/Drawer.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Drawer = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _CloseOutlined = _interopRequireDefault(require("@universe-design/icons-react/CloseOutlined"));
var _isNil = _interopRequireDefault(require("lodash/isNil"));
var _clsx = _interopRequireDefault(require("clsx"));
var _reactTransitionGroup = require("react-transition-group");
var _portal = _interopRequireDefault(require("../portal"));
var _button = require("../button");
var _hooks = require("../config-provider/hooks");
var _shared = require("../shared");
var _util = require("../util");
var _useAnimatedVisible = require("../shared/components/transition/useAnimatedVisible");
var _FadeTransition = require("../shared/components/transition/FadeTransition");
var _SlideTransition = require("../shared/components/transition/SlideTransition");
var _stylex2 = require("../stylex");
require("./style");
/* eslint-disable @typescript-eslint/no-magic-numbers */

/**
 * @public
 */

// TODO: 1. warn 换 utils 内的 2.考虑放到 utils 下
function compatibleDeprecatedProps(props, key, deprecatedKey) {
  if (!(0, _isNil.default)(props[key])) {
    return props[key];
  }
  if (!(0, _isNil.default)(props[deprecatedKey])) {
    console.warn(`"${String(deprecatedKey)}" is deprecated, please use "${String(key)}" instead.`);
    return props[deprecatedKey];
  }
}
const toSlideDirectionMap = {
  left: 'right',
  right: 'left',
  bottom: 'up',
  top: 'down'
};
const useDrawerCssScope = () => (0, _hooks.useCssScope)('drawer');
const getNumericalSize = size => {
  switch (size) {
    case 'large':
      return 680;
    case 'medium':
      return 480;
    case 'small':
      return 350;
    default:
      throw new Error(`Size \`${size}\` is not support.`);
  }
};
const Drawer = props => {
  const {
    visible = false,
    placement = 'right',
    size = 'small',
    width = getNumericalSize(size),
    height = getNumericalSize(size),
    style: wrapperStyle,
    maskStyle,
    mask = true,
    maskClosable = true,
    keyboard = true,
    destroyOnClose = false,
    className: wrapperClassName,
    bodyPaddingSize = 'medium',
    prefixCls,
    zIndex,
    forceRender,
    getContainer,
    afterVisibleChange,
    onClose,
    footer,
    children,
    initialFocusRef,
    finalFocusRef,
    enableFocusLock = false,
    enableRestoreFocus = false
  } = props;
  const showCloseIcon = compatibleDeprecatedProps(props, 'showCloseIcon', 'closable');
  const defaultScope = useDrawerCssScope();
  const cssScope = prefixCls || defaultScope();
  const timeout = _react.default.useMemo(() => ({
    appear: 300,
    enter: 300,
    exit: 200
  }), []);
  const {
    transitionGroupProps,
    transitionGroupPropsWithoutListener,
    styleDisplay,
    animatedVisible,
    noopRef
  } = (0, _useAnimatedVisible.useAnimatedVisible)({
    destroyOnClose,
    forceRender,
    visible,
    onAfterClose: _react.default.useCallback(() => (0, _util.isFunction)(afterVisibleChange) && afterVisibleChange(false), [afterVisibleChange]),
    onAfterOpen: _react.default.useCallback(() => (0, _util.isFunction)(afterVisibleChange) && afterVisibleChange(true), [afterVisibleChange])
  });
  const slideExtraProps = _react.default.useMemo(() => {
    var _ref;
    const direction = toSlideDirectionMap[placement];
    return {
      direction,
      size: (_ref = ['left', 'right'].includes(direction) ? width : height) !== null && _ref !== void 0 ? _ref : 0
    };
  }, [height, placement, width]);
  const renderCloseIcon = () => {
    const {
      closeIcon = /*#__PURE__*/_react.default.createElement(_button.Button, {
        icon: /*#__PURE__*/_react.default.createElement(_CloseOutlined.default, null),
        size: "medium",
        type: "icon"
      })
    } = props;
    return /*#__PURE__*/_react.default.createElement("div", {
      className: `${cssScope}__close`,
      onClick: onClose
    }, closeIcon);
  };
  const renderHeader = () => {
    const {
      title,
      subTitle,
      headerExtra,
      headerStyle
    } = props;
    // compatible other DS's props
    if (!title && !showCloseIcon) {
      return null;
    }
    return /*#__PURE__*/_react.default.createElement("div", {
      className: (0, _clsx.default)(`${cssScope}__header`),
      style: headerStyle
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: `${cssScope}__header__content`
    }, (title || subTitle) && /*#__PURE__*/_react.default.createElement("div", {
      className: (0, _clsx.default)(`${cssScope}__header__title`, (0, _stylex.default)(_stylex2.cssTypographyTitle4))
    }, title), showCloseIcon && renderCloseIcon()), /*#__PURE__*/_react.default.createElement("div", {
      className: `${cssScope}__header__extra`
    }, subTitle && /*#__PURE__*/_react.default.createElement("p", {
      className: (0, _clsx.default)(`${cssScope}__header__subTitle`, (0, _stylex.default)(_stylex2.cssTypographyBody0))
    }, subTitle), headerExtra ? headerExtra : null));
  };
  const renderFooter = () => {
    const {
      footerStyle
    } = props;
    if (!footer) {
      return null;
    }
    return /*#__PURE__*/_react.default.createElement("div", {
      className: `${cssScope}__footer`,
      style: footerStyle
    }, footer);
  };
  const onMaskClick = (0, _react.useCallback)(e => {
    if (!maskClosable) {
      return;
    }
    if (onClose) {
      onClose(e);
    }
  }, [maskClosable, onClose]);
  const maskRef = _react.default.useRef(null);
  const renderMask = () => {
    if (!mask) {
      return;
    }
    const mergeMaskStyle = {
      ...maskStyle,
      ...styleDisplay
    };
    if (zIndex || zIndex === 0) {
      mergeMaskStyle.zIndex = zIndex;
    }
    return /*#__PURE__*/_react.default.createElement(_FadeTransition.FadeTransition, (0, _extends2.default)({
      timeout: timeout,
      appear: true,
      nodeRef: maskRef
    }, transitionGroupPropsWithoutListener), /*#__PURE__*/_react.default.createElement("div", {
      ref: maskRef,
      className: `${cssScope}__mask`,
      style: mergeMaskStyle,
      onClick: onMaskClick,
      tabIndex: -1
    }));
  };
  const onKeyDown = (0, _react.useCallback)(e => {
    if (!keyboard) {
      return;
    }
    // ESC

    if (e.keyCode === 27) {
      e.stopPropagation();
      if (onClose) {
        onClose(e);
      }
    }
  }, [onClose, keyboard]);
  const dialogRef = _react.default.useRef(null);
  const mergedInitialFocusRef = initialFocusRef === null ? null : initialFocusRef !== null && initialFocusRef !== void 0 ? initialFocusRef : dialogRef;
  const renderContent = () => {
    const {
      bodyStyle: bodyStyleProps,
      drawerStyle
    } = props;
    const contentStyle = {
      ...drawerStyle,
      ...styleDisplay,
      ...(height && ['top', 'bottom'].includes(placement) ? {
        height
      } : {}),
      ...(width && ['left', 'right'].includes(placement) ? {
        width
      } : {}),
      zIndex
    };
    const bodyStyle = {
      ...bodyStyleProps,
      ...(typeof bodyPaddingSize === 'number' ? {
        padding: `${bodyPaddingSize}px`
      } : {})
    };
    const bodyPaddingKeywords = typeof bodyPaddingSize !== 'number' ? (0, _util.getDefaultSizeAbbr)(bodyPaddingSize) : '';
    const mergedEnableFocusLock = (() => {
      // 在非 mask 情况, 即使用户没有传 enableFocusLock, 也要关掉lock, 否则按照用户传的为准
      if (!mask && !enableFocusLock) {
        return false;
      }
      return enableFocusLock;
    })();
    return /*#__PURE__*/_react.default.createElement(_SlideTransition.SlideTransition, (0, _extends2.default)({
      appear: true,
      timeout: timeout,
      extraProps: slideExtraProps,
      nodeRef: dialogRef
    }, transitionGroupPropsWithoutListener), /*#__PURE__*/_react.default.createElement(_shared.FocusLock, {
      isDisabled: !visible,
      initialFocusRef: mergedInitialFocusRef,
      restoreFocus: enableRestoreFocus,
      finalFocusRef: finalFocusRef,
      enableFocusLock: mergedEnableFocusLock
    }, /*#__PURE__*/_react.default.createElement("div", {
      ref: dialogRef,
      role: "dialog",
      "aria-modal": "true",
      tabIndex: -1,
      onKeyDown: onKeyDown,
      className: (0, _clsx.default)(`${cssScope}__content__wrapper`),
      style: contentStyle
    }, renderHeader(), /*#__PURE__*/_react.default.createElement("div", {
      className: (0, _clsx.default)(`${cssScope}__content`, {
        [`${cssScope}__content__nofooter`]: footer,
        [`${cssScope}__content__pd-${bodyPaddingKeywords}`]: bodyPaddingKeywords
      }),
      style: bodyStyle
    }, children), renderFooter())));
  };
  return /*#__PURE__*/_react.default.createElement(_reactTransitionGroup.Transition, (0, _extends2.default)({
    nodeRef: noopRef,
    timeout: timeout,
    appear: true
  }, transitionGroupProps), (() => {
    const childNodes = /*#__PURE__*/_react.default.createElement("div", {
      className: (0, _clsx.default)(`${cssScope}__root`, {
        [`${cssScope}-${placement}`]: true,
        [`${cssScope}-open`]: animatedVisible
      }),
      style: wrapperStyle
    }, renderMask(), renderContent());
    if (getContainer === false) {
      return /*#__PURE__*/_react.default.createElement("div", {
        className: wrapperClassName
      }, childNodes);
    }
    return /*#__PURE__*/_react.default.createElement(_portal.default, {
      wrapperClassName: wrapperClassName,
      container: getContainer
    }, childNodes);
  })());
};
exports.Drawer = Drawer;
