// @universe-design/react/src/tree-select/stories/homeless.stories.tsx
import type { Meta } from '@storybook/react';
import remove from 'lodash/remove';
import React, { useEffect, useState } from 'react';
import type { TreeSelectRef } from '../..';
import { Button, Tree, TreeNode, TreeSelect } from '../..';
import type { TreeNodeConfig, TreeProps } from '../../tree/interface';
import { isNonNullable } from '../../util';

export default {
  title: 'Components/TreeSelect/临时示例',
  component: TreeSelect,
} satisfies Meta;

export const TriggerElement = () => (
  <>
    <TreeSelect
      allowClear
      style={{ width: '200px' }}
      placeholder="Select a country"
      dropdownMatchSelectWidth={false}
      triggerElement={<Button>123</Button>}
    >
      <TreeNode value="Parent 1" title="Parent 1">
        <TreeNode value="Parent 1-0" title="Parent 1-0">
          <TreeNode value="leaf1" title="My leaf" />
          <TreeNode value="leaf2" title="Your leaf" />
        </TreeNode>
        <TreeNode value="Parent 1-1" title="Parent 1-1">
          <TreeNode value="sss" title={<b style={{ color: '#3370FF' }}>sss</b>} />
        </TreeNode>
      </TreeNode>
    </TreeSelect>
  </>
);

export function SearchExpand() {
  const data: TreeNodeConfig[] = [
    {
      label: '安道尔',
      value: '6862995746543633928',
      children: [
        {
          label: '安道尔',
          value: '6863326172650931720',
        },
        {
          label: '卡尼略',
          value: '6863326172827092487',
        },
        {
          label: '恩坎普',
          value: '6863326172973893133',
        },
        {
          label: '莱塞斯卡尔德',
          value: '6863326173120693767',
        },
      ],
    },
    {
      label: '安道尔2',
      value: '686299574654363322',
      children: [
        {
          label: '安道尔2',
          value: '68629957465436332',
        },
        {
          label: '卡尼略2',
          value: '68633261728270924872',
        },
        {
          label: '恩坎普2',
          value: '68633261729738931332',
        },
        {
          label: '莱塞斯卡尔德2',
          value: '68633261731206937672',
        },
      ],
    },
  ];

  const [treeExpandedKeys, setTreeExpandedKeys] = useState<string[]>([]);
  const [open, setOpen] = useState(false);
  const [searchValue, setSearchValue] = useState('');
  const [val, setValue] = useState<string | undefined>();

  useEffect(() => {
    if (!open) {
      setTreeExpandedKeys([]);
    }
    const listenerHandle = () => {
      setOpen(!open);
    };
    const item = document.getElementsByClassName('admin-address-tree-select')[0];
    item.addEventListener('click', listenerHandle);
    return () => {
      item.removeEventListener('click', listenerHandle);
    };
  }, [open]);
  console.log('treeExpandedKeys', treeExpandedKeys);
  return (
    <TreeSelect
      className="admin-address-tree-select"
      style={{ width: '200px' }}
      treeData={data}
      showSearch={true}
      placeholder="请选择"
      value={val}
      open={open}
      treeDefaultExpandAll={true}
      searchValue={searchValue}
      onSearch={v => {
        setSearchValue(v);
      }}
      onClear={() => setValue(undefined)}
      // // 失去焦点，收起
      onBlur={() => {
        setOpen(false);
        setSearchValue('');
      }}
      onChange={(selected, meta) => {
        if (!isNonNullable(selected)) {
          return;
        }
        const { option } = meta;
        const children = option?.children;
        // 有子节点，做展开/收起操作
        if (children && children.length > 0) {
          // 收起
          if (treeExpandedKeys.includes(selected)) {
            setTreeExpandedKeys(remove(treeExpandedKeys, item => item !== selected));
          } else {
            setTreeExpandedKeys([...treeExpandedKeys, selected]);
          }
        } else {
          setValue(selected);
          setOpen(false);
          setSearchValue('');
        }
      }}
      onTreeExpand={v => {
        console.log('====', v);
        setTreeExpandedKeys(v);
      }}
      treeExpandedKeys={treeExpandedKeys}
    />
  );
}

const getChildren = (prefix: string) =>
  new Array(20).fill('').map((_, index) => ({
    label: `${prefix} child ${index}`,
    value: `${prefix}_${index}`,
  }));
export const ScrollInfoView = () => {
  const ref = React.useRef<TreeSelectRef>(null);

  const treeData: TreeProps['treeData'] = [
    {
      label: 'Parent 1',
      value: '0-1',
      children: getChildren('1'),
    },
    {
      label: 'Parent 2',
      value: '0-2',
      children: getChildren('2'),
    },
  ];

  return (
    <>
      <Button onClick={() => ref.current?.scrollIntoView({ type: 'node', key: '1_10' }, { align: 'top' })}>
        {' '}
        Scroll To 1_10
      </Button>
      <TreeSelect
        className="admin-address-tree-select"
        style={{ width: '200px' }}
        treeData={treeData}
        showSearch={true}
        placeholder="请选择"
        ref={ref}
        open
      />
    </>
  );
};

export const PerformanceTest = () => {
  const [visible, setVisible] = React.useState(false);

  const root = React.useMemo(() => {
    const getTreeNode = (parentId: string) =>
      new Array(10).fill('').map((i, index) => ({
        value: parentId + '-' + index,
        key: parentId + '-' + index,
        title: parentId + '-' + index,
      }));

    function addChildren(node: TreeNodeConfig, level: number) {
      if (level >= 5) {
        return;
      }
      node.children = getTreeNode(node.value);
      node.children.forEach(childNode => addChildren(childNode, level + 1));
    }
    const _root = getTreeNode('node');
    _root.forEach(childNode => addChildren(childNode, 1));
    console.log(_root);
    return _root;
  }, []);

  return (
    <>
      <button onClick={() => setVisible(!visible)}>toggle tree </button>
      {visible && (
        <Tree
          treeData={root}
          checkable
          // checkStrictly
          // defaultCheckedKeys={["node-0-0"]}
          onCheck={d => console.log(d)}
        />
      )}
    </>
  );
};
