// @universe-design/react/src/tree-select/stories/screenshot.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import CMutiSelectControlledScreenshot from '../demo/MutiSelectControlled.screenshot';
import CMutiSelectControlled2Screenshot from '../demo/MutiSelectControlled2.screenshot';
import CCheckerScreenshot from '../demo/checker.screenshot';
import CLegacy from '../demo/legacy';
import CNoData from '../demo/no-data';
import CZDropdownRender from '../demo/z.dropdown-render';
import CZSearch from '../demo/z.search';
if (!(CMutiSelectControlledScreenshot as Story).storyName) {
  (CMutiSelectControlledScreenshot as Story).storyName = 'Muti select controlled screenshot';
}
if (!(CMutiSelectControlled2Screenshot as Story).storyName) {
  (CMutiSelectControlled2Screenshot as Story).storyName = 'Muti select controlled2 screenshot';
}
if (!(CCheckerScreenshot as Story).storyName) {
  (CCheckerScreenshot as Story).storyName = 'Checker screenshot';
}
if (!(CLegacy as Story).storyName) {
  (CLegacy as Story).storyName = 'Legacy';
}
if (!(CNoData as Story).storyName) {
  (CNoData as Story).storyName = 'No data';
}
if (!(CZDropdownRender as Story).storyName) {
  (CZDropdownRender as Story).storyName = 'Z dropdown render';
}
if (!(CZSearch as Story).storyName) {
  (CZSearch as Story).storyName = 'Z search';
}

export default {
  title: 'Components/TreeSelect/截图测试',
  tags: ['autodocs'],
};

export {
  CMutiSelectControlledScreenshot,
  CMutiSelectControlled2Screenshot,
  CCheckerScreenshot,
  CLegacy,
  CNoData,
  CZDropdownRender,
  CZSearch,
};
