// @universe-design/react/src/tree-select/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Basic from '../demo/0.basic';
import C1Multiple from '../demo/1.multiple';
import C2TreeData from '../demo/2.treeData';
import C3Checkable from '../demo/3.checkable';
import C5Async from '../demo/5.async';
import CDisabledSelectParent from '../demo/disabled-select-parent';
if (!(C0Basic as Story).storyName) {
  (C0Basic as Story).storyName = '0. 基础用法';
  (C0Basic as Story).parameters = {
    docs: {
      description: {
        story: '简单的用法。\n\n',
      },
    },
  };
}
if (!(C1Multiple as Story).storyName) {
  (C1Multiple as Story).storyName = '1. 多选';
  (C1Multiple as Story).parameters = {
    docs: {
      description: {
        story: '多选的树选择。\n\n',
      },
    },
  };
}
if (!(C2TreeData as Story).storyName) {
  (C2TreeData as Story).storyName = '2. 从数据直接生成';
  (C2TreeData as Story).parameters = {
    docs: {
      description: {
        story: '使用`treeData`把 JSON 数据直接生成树结构。\n\n',
      },
    },
  };
}
if (!(C3Checkable as Story).storyName) {
  (C3Checkable as Story).storyName = '3. 可勾选';
  (C3Checkable as Story).parameters = {
    docs: {
      description: {
        story: '使用勾选框实现多选功能。\n\n',
      },
    },
  };
}
if (!(C5Async as Story).storyName) {
  (C5Async as Story).storyName = '5. 异步加载';
  (C5Async as Story).parameters = {
    docs: {
      description: {
        story: '异步加载树节点。\n\n',
      },
    },
  };
}
if (!(CDisabledSelectParent as Story).storyName) {
  (CDisabledSelectParent as Story).storyName = '6. 通过组合的方式禁止选中父节点';
  (CDisabledSelectParent as Story).parameters = {
    docs: {
      description: {
        story: '可以通过组合的方式做到禁止选中父节点但是保留正常的样式\n\n',
      },
    },
  };
}

export default {
  title: 'Components/TreeSelect/官网示例',
  tags: ['autodocs'],
};

export { C0Basic, C1Multiple, C2TreeData, C3Checkable, C5Async, CDisabledSelectParent };
