// @universe-design/rc-table/es/Body/ExpandedRow.js
import * as React from 'react';
import Cell from '../Cell';
import { useTableContextSelector } from '../context/TableContext';
function ExpandedRow(_ref) {
  var prefixCls = _ref.prefixCls,
    children = _ref.children,
    Component = _ref.component,
    cellComponent = _ref.cellComponent,
    fixHeader = _ref.fixHeader,
    fixColumn = _ref.fixColumn,
    horizonScroll = _ref.horizonScroll,
    className = _ref.className,
    expanded = _ref.expanded,
    componentWidth = _ref.componentWidth,
    colSpan = _ref.colSpan;
  var scrollbarColumnWidth = useTableContextSelector(function (ctx) {
    return ctx.scrollbarColumnWidth;
  });
  // Cache render node
  return React.useMemo(function () {
    var contentNode = children;
    if (fixColumn) {
      var width = componentWidth - scrollbarColumnWidth;
      contentNode = /*#__PURE__*/React.createElement("div", {
        style: {
          width: width,
          position: 'sticky',
          left: 0,
          overflow: 'hidden',
          visibility: width > 0 ? undefined : 'hidden' // 防止从宽度为0 到得到宽度时，出现闪烁
        },
        className: "".concat(prefixCls, "-expanded-row-fixed")
      }, contentNode);
    }
    return /*#__PURE__*/React.createElement(Component, {
      className: className,
      style: {
        display: expanded ? null : 'none'
      }
    }, /*#__PURE__*/React.createElement(Cell, {
      component: cellComponent,
      prefixCls: prefixCls,
      colSpan: colSpan
    }, contentNode));
  }, [children, Component, fixHeader, horizonScroll, className, expanded, componentWidth, colSpan, scrollbarColumnWidth]);
}
export default ExpandedRow;
