// @universe-design/rc-table/es/Body/GroupRow.js
import _extends from "@babel/runtime/helpers/esm/extends";
import * as React from 'react';
import Cell from '../Cell';
import { useTableContextSelector } from '../context/TableContext';
import classNames from 'classnames';
export function GroupRow(props) {
  var groupKey = props.groupKey,
    groupRowRenderer = props.groupRowRenderer,
    Component = props.component,
    cellComponent = props.cellComponent,
    colSpan = props.colSpan,
    prefixCls = props.prefixCls,
    triggerExpandGroup = props.triggerExpandGroup,
    onGroupRow = props.onGroupRow,
    componentWidth = props.componentWidth,
    horizonScroll = props.horizonScroll,
    expanded = props.expanded;
  var scrollbarColumnWidth = useTableContextSelector(function (ctx) {
    return ctx.scrollbarColumnWidth;
  });
  // if horizontalScroll we should keep it fixed and set it width equal to table width
  var contentNode = React.useMemo(function () {
    var renderGroupRowTitle = function renderGroupRowTitle() {
      if (typeof groupRowRenderer === 'function') {
        return groupRowRenderer(groupKey, expanded);
      }
      return groupKey;
    };
    return /*#__PURE__*/React.createElement("div", {
      style: {
        width: horizonScroll ? componentWidth - scrollbarColumnWidth : undefined,
        position: 'sticky',
        left: 0
      },
      className: "".concat(prefixCls, "-group-row-content")
    }, renderGroupRowTitle());
  }, [componentWidth, scrollbarColumnWidth, groupKey, groupRowRenderer, expanded, horizonScroll, prefixCls]);
  // Event
  var additionalProps;
  if (onGroupRow) {
    additionalProps = onGroupRow(groupKey);
  }
  var onClick = function onClick(ev) {
    triggerExpandGroup(groupKey);
    if (additionalProps && additionalProps.onClick) {
      var _additionalProps;
      for (var _len = arguments.length, args = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
        args[_key - 1] = arguments[_key];
      }
      (_additionalProps = additionalProps).onClick.apply(_additionalProps, [ev].concat(args));
    }
  };
  var mergedClassName = additionalProps && additionalProps.className ? classNames(additionalProps.className, "".concat(prefixCls, "-group-row")) : "".concat(prefixCls, "-group-row");
  return /*#__PURE__*/React.createElement(Component, _extends({}, additionalProps, {
    className: mergedClassName,
    onClick: onClick
  }), /*#__PURE__*/React.createElement(Cell, {
    component: cellComponent,
    prefixCls: prefixCls,
    className: "".concat(prefixCls, "-group-row-cell"),
    colSpan: colSpan
  }, contentNode));
}
