// @universe-design/rc-table/es/Body/index.js
import _toConsumableArray from "@babel/runtime/helpers/esm/toConsumableArray";
import * as React from 'react';
import BodyRow from './BodyRow';
import { useTableContextSelector } from '../context/TableContext';
import ExpandedRow from './ExpandedRow';
import BodyContext from '../context/BodyContext';
import { getColumnsKey } from '../utils/valueUtil';
import ResizeContext from '../context/ResizeContext';
import MeasureCell from './MeasureCell';
import { GroupRow } from './GroupRow';
function Body(_ref) {
  var data = _ref.data,
    getRowKey = _ref.getRowKey,
    measureColumnWidth = _ref.measureColumnWidth,
    expandedKeys = _ref.expandedKeys,
    onRow = _ref.onRow,
    rowExpandable = _ref.rowExpandable,
    emptyNode = _ref.emptyNode,
    childrenColumnName = _ref.childrenColumnName,
    groupsMap = _ref.groupsMap,
    groupRowRenderer = _ref.groupRowRenderer,
    expandedGroupKeySet = _ref.expandedGroupKeySet,
    triggerExpandGroup = _ref.triggerExpandGroup,
    onGroupRow = _ref.onGroupRow,
    getRecordByKey = _ref.getRecordByKey;
  var _React$useContext = React.useContext(ResizeContext),
    onColumnResize = _React$useContext.onColumnResize;
  var prefixCls = useTableContextSelector(function (ctx) {
    return ctx.prefixCls;
  });
  var getComponent = useTableContextSelector(function (ctx) {
    return ctx.getComponent;
  });
  var _React$useContext2 = React.useContext(BodyContext),
    fixHeader = _React$useContext2.fixHeader,
    horizonScroll = _React$useContext2.horizonScroll,
    flattenColumns = _React$useContext2.flattenColumns,
    componentWidth = _React$useContext2.componentWidth;
  var WrapperComponent = getComponent(['body', 'wrapper'], 'tbody');
  var trComponent = getComponent(['body', 'row'], 'tr');
  var tdComponent = getComponent(['body', 'cell'], 'td');
  return React.useMemo(function () {
    var columnsKey = getColumnsKey(flattenColumns);
    // data row: lines of record
    var renderRows = function renderRows(rowsData) {
      var rows;
      if (rowsData.length) {
        rows = rowsData.map(function (record, index) {
          var key = getRowKey(record, index);
          // one record row and expanded row (if expandable)
          return /*#__PURE__*/React.createElement(BodyRow, {
            key: key,
            rowKey: key,
            record: record,
            recordKey: key,
            index: index,
            rowComponent: trComponent,
            cellComponent: tdComponent,
            expandedKeys: expandedKeys,
            onRow: onRow,
            getRowKey: getRowKey,
            rowExpandable: rowExpandable,
            childrenColumnName: childrenColumnName
          });
        });
      } else {
        rows = /*#__PURE__*/React.createElement(ExpandedRow, {
          expanded: true,
          className: "".concat(prefixCls, "-placeholder"),
          prefixCls: prefixCls,
          fixHeader: fixHeader,
          fixColumn: horizonScroll,
          horizonScroll: horizonScroll,
          component: trComponent,
          componentWidth: componentWidth,
          cellComponent: tdComponent,
          colSpan: flattenColumns.length
        }, emptyNode);
      }
      return rows;
    };
    // all table rows, contains group title、data row、expanded row
    var renderTableRows = function renderTableRows() {
      var groupKeys = _toConsumableArray(groupsMap.keys());
      // grouped data
      if (!!groupKeys.length) {
        var groups = groupKeys.map(function (key) {
          var groupExpanded = expandedGroupKeySet.has(key);
          // group title row
          var renderGroupRow = function renderGroupRow() {
            return /*#__PURE__*/React.createElement(GroupRow, {
              groupKey: key,
              prefixCls: prefixCls,
              groupRowRenderer: groupRowRenderer,
              triggerExpandGroup: triggerExpandGroup,
              onGroupRow: onGroupRow,
              component: trComponent,
              cellComponent: tdComponent,
              colSpan: flattenColumns.length,
              componentWidth: componentWidth,
              horizonScroll: horizonScroll,
              expanded: groupExpanded
            });
          };
          //group data rows
          var renderGroupedDataRows = function renderGroupedDataRows() {
            var groupDataKeys = groupsMap.get(key);
            var groupData = groupDataKeys.map(function (k) {
              return getRecordByKey(k);
            });
            return renderRows(groupData);
          };
          return /*#__PURE__*/React.createElement(React.Fragment, {
            key: key
          }, renderGroupRow(), groupExpanded && renderGroupedDataRows());
        });
        return groups;
      } else {
        return renderRows(data);
      }
    };
    return /*#__PURE__*/React.createElement(WrapperComponent, {
      className: "".concat(prefixCls, "-tbody")
    }, measureColumnWidth && /*#__PURE__*/React.createElement("tr", {
      "aria-hidden": "true",
      className: "".concat(prefixCls, "-measure-row"),
      style: {
        height: 0,
        fontSize: 0
      }
    }, columnsKey.map(function (columnKey) {
      return /*#__PURE__*/React.createElement(MeasureCell, {
        key: columnKey,
        columnKey: columnKey,
        onColumnResize: onColumnResize
      });
    })), renderTableRows());
  }, [data, prefixCls, emptyNode, expandedGroupKeySet, expandedKeys, flattenColumns, getRowKey, measureColumnWidth, onRow, componentWidth, WrapperComponent, childrenColumnName, fixHeader, getRecordByKey, groupRowRenderer, groupsMap, horizonScroll, onColumnResize, onGroupRow, rowExpandable, tdComponent, trComponent, triggerExpandGroup]);
}
var MemoBody = /*#__PURE__*/React.memo(Body);
MemoBody.displayName = 'Body';
export default MemoBody;
