// @universe-design/react/src/date-picker/RangePicker/RangePickerPanel.tsx
import React, { useState } from 'react';
import type moment from 'moment';
import clsx from 'clsx';

import { Space } from '../../space';
import type { MODE } from '../interface';
import { getHigherLevelMode, getNext, getPrev } from '../panels/base/utils';
import { useRangePickerContext } from './Context';
import type { RangePickerPanelProps } from './types';
import RangePickerHeader from './RangePickerHeader';
import {
  NextMonthSwitcher,
  NextYearSwitcher,
  Next20YearsSwitcher,
  PreMonthSwitcher,
  PreYearSwitcher,
  Pre20YearsSwitcher,
} from '../panels/PanelHeaderSwitchers';
import type { MonthYearSelectorProps } from '../panels/selector/MonthYearSelector';
import { MonthYearSelector } from '../panels/selector/MonthYearSelector';
import stylex from '@universe-design/stylex';
import { sxPickerHeaderSwitchWrapper } from '../sx';

import { DateCalendarUI } from '../panels/DateCalendar/DateCalendarUI';
import type { DateCalendarUIProps } from '../panels/DateCalendar/interface';
import '../style';

/**
 * 多层选择器解决方案:
 * 当前mode下面，选择日期，才能生效
 * 其他mode上面选择日期，都只是一个中间态，可以在panel组件内部解决
 */

function getChanged(date: moment.Moment, mode: MODE) {
  const seq: MODE[] = ['date', 'month', 'year'];
  const index = seq.indexOf(mode);
  if (index < 0) {
    return {};
  }
  return seq.slice(index).reduce((pre: { [_: string]: number }, mod: MODE) => {
    pre[mod] = date.get(mod);
    return pre;
  }, {});
}

function RangePickerPanel(
  props: RangePickerPanelProps & {
    className?: string;
  },
) {
  // Props
  const { className, dateRender, startOfWeek, hideWeekNumbers, janX } = props;

  // Context
  const {
    prefixCls,
    viewValue,
    mode,
    handleDateSelect,
    handleDateCellMouseEnter,
    handleDateCellMouseLeave,
    isDisabled,
    isPreviewDate,
    localMode,
    setLocalMode,
    startPickerValue,
    endPickerValue,
    setStartPickerValue,
    setEndPickerValue,
    cssScope,
    panelIntervalLimits,
    maxDate,
    minDate,
  } = useRangePickerContext();
  const panelLimited = panelIntervalLimits === 'adjacent';

  const [focusedIndex, setFocusedIndex] = useState(0);

  const handleFocusedPanelSelect = (date: moment.Moment) => {
    const newViewDate = date.clone();
    newViewDate.set(getChanged(date, localMode));
    if (focusedIndex === 1) {
      setEndPickerValue(newViewDate);
    } else {
      setStartPickerValue(newViewDate);
    }

    if (mode !== localMode) {
      setLocalMode(mode);
    }
  };

  const handleCollapsed = (index: number) => {
    setFocusedIndex(index);
    const nextMode = getHigherLevelMode(mode);
    if (nextMode) {
      setLocalMode(nextMode);
    }
  };

  const handleMonthClick = (index: number) => {
    setFocusedIndex(index);
    setLocalMode('month');
  };

  const handleYearClick = (index: number) => {
    setFocusedIndex(index);
    setLocalMode('year');
  };

  if (localMode !== mode) {
    return (
      <DateCalendarUI
        sx={[
          stylex.create({
            margin: 15,
          }),
        ]}
        key={focusedIndex}
        mode={localMode}
        disabledDate={isDisabled}
        defaultPickerValue={focusedIndex === 0 ? startPickerValue : endPickerValue}
        onSelect={handleFocusedPanelSelect}
        startOfWeek={startOfWeek}
        minDate={minDate}
        maxDate={maxDate}
        hideWeekNumbers={hideWeekNumbers}
        janX={janX}
      />
    );
  }

  const getPickerBodyProps = (index: 0 | 1): DateCalendarUIProps => {
    const first = index === 0;
    const dates = viewValue.filter(x => Boolean(x)) as moment.Moment[];
    return {
      disabledDate: isDisabled,
      minDate,
      maxDate,
      previewDate: isPreviewDate,
      pickerValue: first ? startPickerValue : endPickerValue,
      mode,
      onSelect: handleDateSelect,
      onDateMouseEnter: handleDateCellMouseEnter,
      onContentMouseLeave: handleDateCellMouseLeave,

      startOfWeek,
      hideWeekNumbers,
      janX,
      ...(dateRender
        ? { cellRenderer: (date, { viewMode, defaultNode }) => dateRender(date, viewMode, defaultNode) }
        : {}),
      ...(viewValue[0] && viewValue[1]
        ? {
            selectedRange: viewValue as [moment.Moment, moment.Moment],
          }
        : { selectedDates: dates }),
    };
  };

  const getMonthYearSelectorProps = (index: 0 | 1): MonthYearSelectorProps => {
    const MonthYearSelectorCls = cssScope('-panel-header-content__center');
    const resValue = index === 0 ? startPickerValue : endPickerValue;
    return {
      mode,
      pickerValue: resValue,
      className: MonthYearSelectorCls,
      hiddenCollapse: mode === 'year',
      collapsed: localMode === mode,
      onMonthClick: () => handleMonthClick(index),
      onYearClick: () => handleYearClick(index),
      onCollapsed: () => handleCollapsed(index),
    };
  };

  const isDate = mode === 'date';
  const isYear = mode === 'year';

  // TODO handleFns
  const pickerPanelRenderer = (index: 0 | 1) => {
    const isLeft = index === 0;
    const isRight = index === 1;
    const pickerValue = isLeft ? startPickerValue.clone() : endPickerValue.clone();
    const switchPicker = (value: moment.Moment) => {
      if (isLeft) {
        setStartPickerValue(value.clone());
      } else {
        setEndPickerValue(value.clone());
      }
    };

    // handleTimeFns
    const goPrevMonth = () => switchPicker(getPrev(pickerValue, 'date'));
    const goNextMonth = () => switchPicker(getNext(pickerValue, 'date'));
    const goPrevYear = () => switchPicker(getPrev(pickerValue, 'month'));
    const goNextYear = () => switchPicker(getNext(pickerValue, 'month'));
    const goPre20Years = () => switchPicker(getPrev(pickerValue, 'year'));
    const goNext20Years = () => switchPicker(getNext(pickerValue, 'year'));

    return (
      <DateCalendarUI
        headerRenderer={() => (
          <RangePickerHeader>
            <Space size={4} className={stylex(sxPickerHeaderSwitchWrapper)}>
              {isYear && <Pre20YearsSwitcher asPlaceholder={isRight && panelLimited} onClick={goPre20Years} />}
              {!isYear && <PreYearSwitcher asPlaceholder={isRight && panelLimited} onClick={goPrevYear} />}
              {isDate && <PreMonthSwitcher asPlaceholder={isRight && panelLimited} onClick={goPrevMonth} />}
            </Space>
            <MonthYearSelector {...getMonthYearSelectorProps(index)} />
            <Space size={4} className={stylex(sxPickerHeaderSwitchWrapper)}>
              {isDate && <NextMonthSwitcher asPlaceholder={isLeft && panelLimited} onClick={goNextMonth} />}
              {!isYear && <NextYearSwitcher asPlaceholder={isLeft && panelLimited} onClick={goNextYear} />}
              {isYear && <Next20YearsSwitcher asPlaceholder={isLeft && panelLimited} onClick={goNext20Years} />}
            </Space>
          </RangePickerHeader>
        )}
        {...getPickerBodyProps(index)}
      />
    );
  };

  return (
    <>
      <div className={clsx(className, `${prefixCls}-panels`)}>
        {pickerPanelRenderer(0)}
        {pickerPanelRenderer(1)}
      </div>
    </>
  );
}

export { RangePickerPanel };
