// @universe-design/react/src/date-picker/RangePicker/range-picker-strategies.ts
import type React from 'react';
import type moment from 'moment';
import type { RangeValue } from '../interface';
import { Phase, TriggerType } from './types';

export interface RangePickerSelectionStrategyContext {
  // 更新当前选取阶段
  setPhase: (value: React.SetStateAction<Phase>) => void;
  // 当前阶段
  phaseRef: React.MutableRefObject<Phase>;
  // 提交变更给用户
  commitChange: () => void;
  // 打开关闭 overlay
  setOpened: (visible: boolean, triggerType?: TriggerType) => void;
  // 更新 view value，这个值是暂时值，改变不会直接提交给用户
  setViewValue: (
    value: React.SetStateAction<RangeValue>,
    options?: {
      /**
       * 是否自动提交结果，提交结果会触发 onChange 将结果提交给用户
       */
      autoCommit?: boolean;
    },
  ) => void;
  // 当前的 view value
  viewValueRef: React.MutableRefObject<RangeValue>;
  // 选取模式，年 | 月 | 日
  mode: 'date' | 'month' | 'year';
}

export interface RangePickerSelectionStrategy {
  /**
   * 策略名称
   */
  name?: string;
  /**
   * 默认阶段,开始时间或结束时间
   */
  defaultPhase: Phase;
  /**
   * 相关事件处理
   */
  useHooks: (context: RangePickerSelectionStrategyContext) => {
    /**
     * @returns 本次选择日期所属的阶段（开始时间、结束时间)
     */
    onDateSelect: (date: moment.Moment) => Phase | void;
    onPopupVisibleChange: (visible: boolean) => void;
  };
}

const rangePickerSelectionStrategyStartDate: RangePickerSelectionStrategy = {
  name: 'StartDate',
  defaultPhase: Phase.StartDate,
  useHooks: ({ phaseRef, viewValueRef, mode, setPhase, setViewValue, setOpened }) => {
    const onDateSelect = (date: moment.Moment) => {
      const currentPhase = phaseRef.current;
      const [start, end] = viewValueRef.current;
      // 选择开始时间
      if (currentPhase === Phase.StartDate) {
        // 需要重新选取结束时间的情况
        // 1. 如果选择时间在结束时间之后，说明结束时间失效了
        // 2. 如果没有选取过结束时间
        if (date.isAfter(end, mode) || end === undefined) {
          setPhase(Phase.EndDate);
          setViewValue([date, undefined]);
          return;
        }
        // 对于合法的情况，直接填入选择值，并更新给用户。
        setViewValue([date, end], { autoCommit: true });
        setOpened(false);
        return;
      }
      // 选择结束时间
      if (currentPhase === Phase.EndDate) {
        // 如果选中的结束时间比开始时间早就认为是重置开始时间
        if (date.isBefore(start, mode)) {
          setPhase(Phase.EndDate);
          setViewValue([date, undefined]);
          return Phase.StartDate;
        }
        // 选择完毕以后直接提交退出
        setViewValue([start, date], { autoCommit: true });
        setOpened(false);
        return;
      }
    };

    const onPopupVisibleChange = (visible: boolean) => {
      if (visible) {
        setPhase(Phase.StartDate);
        setOpened(true, TriggerType.StartDate);
      } else {
        setOpened(false);
      }
    };

    return {
      onDateSelect,
      onPopupVisibleChange,
    };
  },
};

const rangePickerSelectionStrategyEndDate: RangePickerSelectionStrategy = {
  name: 'EndDate',
  defaultPhase: Phase.EndDate,
  useHooks: ({ setPhase, phaseRef, viewValueRef, setViewValue, mode, setOpened }) => {
    const onDateSelect = (date: moment.Moment) => {
      const currentPhase = phaseRef.current;
      const [start, end] = viewValueRef.current;
      if (currentPhase === Phase.EndDate) {
        if (date.isBefore(start, mode) || start === undefined) {
          setPhase(Phase.StartDate);
          setViewValue([undefined, date]);
          return;
        }
        setViewValue([start, date], { autoCommit: true });
        setOpened(false);
        return;
      }
      if (currentPhase === Phase.StartDate) {
        if (date.isAfter(end, mode)) {
          setPhase(Phase.StartDate);
          setViewValue([undefined, date]);
          return Phase.EndDate;
        }
        setViewValue([date, end], { autoCommit: true });
        setOpened(false);
        return;
      }
    };

    const onPopupVisibleChange = (visible: boolean) => {
      if (visible) {
        setPhase(Phase.EndDate);
        setOpened(true, TriggerType.EndDate);
      } else {
        setOpened(false);
      }
    };

    return {
      onDateSelect,
      onPopupVisibleChange,
    };
  },
};

const rangePickerSelectionStrategyStartDateOverall: RangePickerSelectionStrategy = {
  name: 'StartDateOverall',
  defaultPhase: Phase.StartDate,
  useHooks: ({ setPhase, phaseRef, viewValueRef, setViewValue, mode, setOpened }) => {
    const onDateSelect = (date: moment.Moment) => {
      /**
       * 规则
       * 1. 只要是合法的值就 commit
       * 2. 每次都完整进行 start -> end 的操作
       */
      const currentPhase = phaseRef.current;
      const [start, end] = viewValueRef.current;
      if (currentPhase === Phase.StartDate) {
        if (date.isAfter(end, mode) || end === undefined) {
          // 不合法直接取消之前的结束时间
          setPhase(Phase.EndDate);
          setViewValue([date, undefined]);
          return;
        }
        setPhase(Phase.EndDate);
        setViewValue([date, end], { autoCommit: true });
        return;
      }

      if (currentPhase === Phase.EndDate) {
        if (!start) {
          setPhase(Phase.StartDate);
          setViewValue([undefined, date]);
          return;
        }
        if (date.isBefore(start, mode)) {
          setPhase(Phase.EndDate);
          setViewValue([date, undefined]);
          return Phase.StartDate;
        }
        // 提交给用户
        setPhase(Phase.StartDate);
        setViewValue([start, date], { autoCommit: true });
        setOpened(false);
        return;
      }
    };

    const onPopupVisibleChange = (visible: boolean) => {
      if (visible) {
        setPhase(Phase.StartDate);
        setOpened(true);
      } else {
        setOpened(false);
      }
    };

    return {
      onDateSelect,
      onPopupVisibleChange,
    };
  },
};

export {
  rangePickerSelectionStrategyStartDate,
  rangePickerSelectionStrategyEndDate,
  rangePickerSelectionStrategyStartDateOverall,
};
