// @universe-design/react/src/date-picker/RangePicker/types.ts
import type React from 'react';
import type { TriggerProps } from '@universe-design/rc-trigger';
import type moment from 'moment';
import type { IFormat, IParse, MODE, RangeValue, PickerType, StartOfWeek } from '../interface';
import type { DefaultSizeType } from '../../util';
import type { RangePickerContextType } from './Context';
import type { RangePickerSelectionStrategy } from './range-picker-strategies';
import type { DateCalendarUIProps } from '../panels/DateCalendar/interface';

export enum Phase {
  StartDate,
  EndDate,
}

export enum TriggerType {
  None,
  Unknown,
  StartDate,
  EndDate,
}

export interface RangePickerState {
  startText?: string;
  endText?: string;
  startHoverText?: string;
  endHoverText?: string;
  value?: RangeValue;
  open?: boolean;
}

export type RangePickerTriggerElement = React.ReactNode | ((state: RangePickerState) => React.ReactNode);

interface RangePickerSharedProps {
  triggerElement?: RangePickerTriggerElement;
}
export interface RangePickerTriggerProps extends RangePickerSharedProps {
  /**
   * 输入框大小
   * @defaultValue `medium`
   */
  size?: DefaultSizeType;
  /**
   * 是否是错误状态
   * @defaultValue false
   */
  error?: boolean;
  /**
   * 是否显示清除按钮
   * @defaultValue true
   */
  allowClear?: boolean;
  /**
   * `className` 属性透传
   */
  className?: string;
  /**
   * `style` 属性透传
   */
  style?: React.CSSProperties;
  /**
   * 自定义的选择框后缀图标
   */
  suffixIcon?: React.ReactNode;
  /**
   * 原生事件
   */
  onMouseDown?: React.MouseEventHandler<HTMLDivElement>;
  /**
   * 原生事件
   */
  onClick?: React.MouseEventHandler<HTMLDivElement>;
  /**
   * 弹层元素的 `className` 属性透传
   */
  dropdownClassName?: string;
  /**
   * 定义浮层的容器，默认为 body 上新建 div
   */
  getPopupContainer?: TriggerProps['getPopupContainer'];
  /**
   * 弹层元素的 `style` 属性透传
   */
  popupStyle?: React.CSSProperties;
  /**
   * 输入框提示文字
   */
  placeholder?: [string, string];
  /**
   * 原生事件
   */
  onMouseUp?: React.MouseEventHandler<HTMLDivElement>;
  /**
   * 原生事件
   */
  onKeyDown?: React.KeyboardEventHandler<HTMLInputElement>;
  /**
   * 设置输入框为只读（避免在移动设备上打开虚拟键盘）
   * @defaultValue false
   */
  inputReadOnly?: boolean;
  /**
   * 自动获取焦点
   * @defaultValue false
   */
  autoFocus?: boolean;
}

/**
 * @public
 */
export interface RangePickerPanelProps {
  /**
   * 自定义日期单元格的内容
   */
  dateRender?: (currentDate: moment.Moment, viewMode: MODE, defaultNode: React.ReactNode) => React.ReactNode;
  /**
   * 一周开始于星期几，默认星期天
   * @defaultValue 0
   */
  startOfWeek?: StartOfWeek;
  /**
   * 是否隐藏周数
   * @defaultValue true
   */
  hideWeekNumbers?: boolean;
  /**
   * 一年的第一周需要包含1月的 X 天，见 [Moment](https://momentjs.cn/docs/#/customization/dow-doy/)
   */
  janX?: number;
}

export interface RangePickerFooterProps {
  /**
   * 在面板中添加额外的页脚
   */
  renderExtraFooter?: (ctx: RangePickerContextType) => React.ReactNode;
  /**
   * 预设时间范围快捷选择
   */
  ranges?: {
    [range: string]: [moment.Moment, moment.Moment] | (() => [moment.Moment, moment.Moment]);
  };
}

export interface RangePickerContextProps
  extends RangePickerSharedProps,
    Pick<DateCalendarUIProps, 'minDate' | 'maxDate' | 'mode' | 'disabledDate'> {
  /**
   * @defaultValue `date`
   * @deprecated 请使用 mode
   */
  picker?: PickerType;
  /**
   * 选中的日期的受控值
   */
  value?: RangeValue;
  /**
   * 选中的日期的默认值
   */
  defaultValue?: RangeValue;
  /**
   * 选中日期值变化的回调
   */
  onChange?: (value: RangeValue) => void;
  /**
   * 自定义解析 input 字符串的方法
   */
  parse?: IParse;
  /**
   * 日期显示格式，见 [Moment](https://momentjs.com/docs/#/displaying/)
   */
  format?: IFormat;
  /**
   * 控制弹层是否展开的受控值
   */
  open?: boolean;
  /**
   * 控制弹层是否展开的默认值
   */
  defaultOpen?: boolean;
  /**
   * 禁用选日期范围
   * @defaultValue false
   */
  disabled?: [boolean?, boolean?];
  /**
   * 弹出日历和关闭日历的回调
   */
  onOpenChange?: (open: boolean) => void;
  /**
   * 点击选中日期的回调
   */
  onSelect?: (value: moment.Moment, phase: Phase) => void;
  /**
   * 点击清除按钮时的回调
   */
  onClear?: () => void;
  /**
   * 自定义选中交互策略，默认使用 `autoCommit` 提交
   */
  selectionStrategy?: RangePickerSelectionStrategy;
  /**
   * 当 value 为 null 时，面板显示的日期
   */
  defaultPickerValue?: [moment.Moment, moment.Moment?];
  /**
   * 左右面板切换联动限制，当设置为 `none` 时，取消联动限制
   * @defaultValue 'adjacent'
   */
  panelIntervalLimits?: 'none' | 'adjacent';
}

/**
 * @public
 * @docgen
 */
export interface RangePickerProps
  extends RangePickerTriggerProps,
    RangePickerContextProps,
    RangePickerPanelProps,
    RangePickerFooterProps {}
