// @universe-design/react/src/dropdown/stories/homeless.stories.tsx
import * as React from 'react';
import { Dropdown, Space, toast, Divider, Button, Tooltip, Checkbox } from '../..';
import { Menu } from '../../menu';
import InfoOutlined from '@universe-design/icons-react/InfoOutlined';
import MemberOutlined from '@universe-design/icons-react/MemberOutlined';
import SpaceDownBoldOutlined from '@universe-design/icons-react/SpaceDownBoldOutlined';
import { RemoveScroll } from 'react-remove-scroll';

import './index.less';

export default {
  title: 'Components/Dropdown/临时示例',
};

const link = 'https://universe.bytedance.net/';
const menu = (
  <Menu selectable onSelect={e => console.log(1, e)} onClick={e => console.log(2, e)} selectedKeys={['5.1', '6']}>
    <Menu.Item icon={<InfoOutlined />} onClick={() => toast.info('menu item click')}>
      <a target="_blank" rel="noopener noreferrer" href={link}>
        1st menu item
      </a>
    </Menu.Item>
    <Menu.Item>
      <a target="_blank" rel="noopener noreferrer" href={link}>
        2nd menu item
      </a>
    </Menu.Item>
    <Menu.Item icon={<InfoOutlined />} disabled>
      3rd menu item
    </Menu.Item>
    <Menu.SubMenu disabled title="4rd menu item 4rd menu item" icon={<InfoOutlined />}>
      <Menu.Item>1</Menu.Item>
      <Menu.Item>2</Menu.Item>
      <Menu.Item>3</Menu.Item>
    </Menu.SubMenu>
    <Menu.ItemGroup title="13">
      <Menu.Item>1</Menu.Item>
      <Menu.SubMenu title="4rd menu item">
        <Menu.Item>1</Menu.Item>
        <Menu.Item>2</Menu.Item>
        <Menu.Item>3</Menu.Item>
        <Menu.SubMenu title="5rd menu item">
          <Menu.Item>1</Menu.Item>
          <Menu.Item>2</Menu.Item>
          <Menu.Item>3</Menu.Item>
          <Menu.Item>4</Menu.Item>
          <Menu.Item>5</Menu.Item>
          <Menu.SubMenu title="10rd menu item">
            <Menu.Item>1</Menu.Item>
            <Menu.Item>2</Menu.Item>
            <Menu.Item>3</Menu.Item>
          </Menu.SubMenu>
        </Menu.SubMenu>
      </Menu.SubMenu>
      <Menu.Divider />
      <Menu.Item>1</Menu.Item>
    </Menu.ItemGroup>
    <Menu.SubMenu title="5rd menu item">
      <Menu.Item key="5.1">1</Menu.Item>
      <Menu.Item>2</Menu.Item>
      <Menu.Item>3</Menu.Item>
      <Menu.SubMenu disabled title="4rd menu item">
        <Menu.Item>1</Menu.Item>
        <Menu.Item>2</Menu.Item>
        <Menu.Item>3</Menu.Item>
      </Menu.SubMenu>
      <Menu.Divider />
      <Menu.Item>4</Menu.Item>
    </Menu.SubMenu>
    <Menu.Item key="6" icon={<InfoOutlined />}>
      6rd menu item
    </Menu.Item>
  </Menu>
);

const menu2 = (
  <Menu selectable={false}>
    <Menu.Item>123</Menu.Item>
  </Menu>
);

export const AllDropdown = () => (
  <div
    style={{
      height: 2000,
    }}
    className="ud-dropdown-demo"
  >
    <Divider text="Dropdown" />
    <div>
      <Dropdown overlay={menu} trigger={['click']}>
        <span>点我渲染</span>
      </Dropdown>
      <br />
      <br />
      <Dropdown overlay={menu} trigger={['click']}>
        <Button>点我渲染</Button>
      </Dropdown>
      <br />
      <br />
      <Dropdown disabled overlay={menu} trigger={['click']}>
        <Button>点我渲染 - 禁用</Button>
      </Dropdown>
      <Dropdown.Button disabled overlay={menu} icon={<SpaceDownBoldOutlined className="custom-icon" />}>
        <span data-testid="left-button">button - 禁用</span>
      </Dropdown.Button>
    </div>
    <Divider text="Dropdown button" />
    <div style={{ marginLeft: 100 }}>
      <Dropdown.Button overlay={menu}>Hello</Dropdown.Button>
      <Dropdown.Button type="primary" overlay={menu} placement="bottomCenter">
        Hello
      </Dropdown.Button>
      <Dropdown.Button overlay={menu} placement="bottomLeft">
        Hello
      </Dropdown.Button>
      <Dropdown.Button overlay={menu} icon={<MemberOutlined />}>
        Hello
      </Dropdown.Button>
      <Dropdown.Button
        overlay={menu}
        buttonsRender={([leftButton, rightButton]) => [
          <Tooltip title="tooltip" key="leftButton">
            {leftButton}
          </Tooltip>,
          React.cloneElement(rightButton as React.ReactElement, {
            loading: true,
          }),
        ]}
      >
        hello
      </Dropdown.Button>
    </div>
    <Divider text="事件" />
    <div>
      <Dropdown overlay={menu}>
        <Button>hover</Button>
      </Dropdown>
      <Dropdown overlay={menu} trigger="click">
        <Button>click</Button>
      </Dropdown>
      <Dropdown overlay={menu} trigger="contextMenu">
        <Button>contextmenu</Button>
      </Dropdown>
      <Dropdown overlay={menu} trigger={['click', 'hover']}>
        <Button>hover、click</Button>
      </Dropdown>
    </div>
    <Divider text="Dropdown 弹出位置" />
    <div style={{ textAlign: 'center' }}>
      <Dropdown
        overlay={menu}
        placement="bottomLeft"
        trigger={['click']}
        align={{
          overflow: {
            adjustY: false,
          },
        }}
      >
        <Button>bottom - left</Button>
      </Dropdown>
      <Dropdown overlay={menu} placement="bottomCenter" trigger={['click']}>
        <Button>bottom - center</Button>
      </Dropdown>
      <Dropdown overlay={menu} placement="bottomRight" trigger={['click']}>
        <Button>bottom - right</Button>
      </Dropdown>
    </div>
    <div style={{ textAlign: 'center' }}>
      <Dropdown overlay={menu} placement="topLeft" trigger={['click']}>
        <Button>top - left</Button>
      </Dropdown>
      <Dropdown overlay={menu} placement="topCenter" trigger={['click']}>
        <Button>top - center</Button>
      </Dropdown>
      <Dropdown overlay={menu} placement="topRight" trigger={['click']}>
        <Button>top - right</Button>
      </Dropdown>
    </div>
    <Divider text="Menu 配置" />
    <div>
      <Dropdown overlay={menu2} placement="topLeft" trigger={['click']}>
        <Button>menu - selectable: false</Button>
      </Dropdown>
    </div>
  </div>
);

export const CustomStyle = () => (
  <>
    <div className="ud-dropdown-demo">
      <Dropdown
        overlay={menu}
        // maxHeight={300}
        trigger={['click']}
      >
        <span>点我渲染</span>
      </Dropdown>
    </div>
  </>
);

export { default as TextOverflow } from '../demo/text-overflow';

export const PlacementDemo = () => {
  const placementMenu = (
    <Menu>
      <Menu.Item>
        <a target="_blank" rel="noopener noreferrer" href={link}>
          1st menu item
        </a>
      </Menu.Item>
      <Menu.Item>
        <a target="_blank" rel="noopener noreferrer" href={link}>
          2nd menu item
        </a>
      </Menu.Item>
      <Menu.Item>
        <a target="_blank" rel="noopener noreferrer" href={link}>
          3rd menu item
        </a>
      </Menu.Item>
    </Menu>
  );

  return (
    <Space size="large" style={{ marginTop: 200, width: '100%', justifyContent: 'center' }}>
      <Dropdown overlay={placementMenu} placement="topCenter">
        <Button>topCenter</Button>
      </Dropdown>
      <Dropdown overlay={placementMenu} placement="topLeft">
        <Button>topLeft</Button>
      </Dropdown>
      <Dropdown overlay={placementMenu} placement="topRight">
        <Button>topRight</Button>
      </Dropdown>
      <Dropdown overlay={placementMenu} placement="bottomCenter">
        <Button>bottomCenter</Button>
      </Dropdown>
      <Dropdown overlay={placementMenu} placement="bottomLeft">
        <Button>bottomLeft</Button>
      </Dropdown>
      <Dropdown overlay={placementMenu} placement="bottomRight">
        <Button>bottomRight</Button>
      </Dropdown>
    </Space>
  );
};

export const visibleChangeDemo = () => {
  // eslint-disable-next-line react-hooks/rules-of-hooks
  const [visible, setVisible] = React.useState<boolean>(false);

  const onVisibleChange = (v: boolean) => {
    console.log('visible', v);
    setVisible(v);
  };

  return (
    <>
      <Button onClick={() => setVisible(!visible)}>{visible ? '关' : '开'}</Button>
      <Dropdown
        overlay={menu}
        // visible={visible}
        onVisibleChange={onVisibleChange}
        placement="bottomRight"
      >
        <Button>bottomRight</Button>
      </Dropdown>
    </>
  );
};

export const AlignDemo = () => (
  <Dropdown overlay={menu} placement="bottomRight" align={{ offset: [0, 8] }}>
    <Button>bottomRight</Button>
  </Dropdown>
);

export const Large = () => (
  <Dropdown trigger="click" overlay={menu} placement="bottomCenter">
    <Button style={{ marginLeft: 'auto' }}>bottomRight</Button>
  </Dropdown>
);

export const Multiple = () => {
  const [selectedKeys, setSelectedKeys] = React.useState<string[]>([]);
  const update = (type: string, v: string[]) => {
    console.log(type, v);
    setSelectedKeys(v);
  };
  const MultipleMenu = (
    <Menu
      mode="vertical"
      multiple
      selectable
      selectedKeys={selectedKeys}
      onSelect={info => update('onSelect', info.selectedKeys as any)}
      onDeselect={info => update('onDeselect', info.selectedKeys as any)}
    >
      <Menu.Item>1</Menu.Item>
      <Menu.Item>2</Menu.Item>
      <Menu.Item>3</Menu.Item>
    </Menu>
  );
  return (
    <Dropdown trigger="click" overlay={MultipleMenu} placement="bottomCenter">
      <Button style={{ marginLeft: 'auto' }}>bottomRight</Button>
    </Dropdown>
  );
};

export const FullPageContextMenu = () => (
  <Dropdown trigger="contextMenu" overlay={menu} placement="bottomCenter">
    <div className="ud-dropdown-demo-full-page">
      <div>page</div>
    </div>
  </Dropdown>
);

export const CheckBox = () => {
  const checkboxmenu = (
    <Menu onClick={() => console.trace('onClick')}>
      <Menu.Item>
        <Checkbox>确定</Checkbox>
      </Menu.Item>
    </Menu>
  );

  return (
    <Dropdown trigger="click" overlay={checkboxmenu} placement="bottomCenter">
      <Button>点我</Button>
    </Dropdown>
  );
};

export const PreventScroll = () => {
  const [visible, setVisible] = React.useState(false);
  return (
    <div>
      <div style={{ height: 400 }} />
      <Dropdown
        onVisibleChange={setVisible}
        overlay={
          <RemoveScroll removeScrollBar={false} enabled={visible}>
            {menu}
          </RemoveScroll>
        }
      >
        <Button type="primary">Hover me</Button>
      </Dropdown>
      <div style={{ height: 1200 }} />
    </div>
  );
};

export { default as NormalMenu } from '../demo/all';
export { default as NormalMenuHideCheckmark } from '../demo/all.hideCheckmark';
