// @universe-design/react/src/dropdown/stories/site.stories.tsx
// autogen by ../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import CBasic from '../demo/basic';
import CButton from '../demo/button';
import CContextMenu from '../demo/context-menu';
import CEvent from '../demo/event';
import CItem from '../demo/item';
import COverlayVisible from '../demo/overlay-visible';
import CPlacement from '../demo/placement';
import CSubMenu from '../demo/sub-menu';
import CTrigger from '../demo/trigger';
if (!(CBasic as Story).storyName) {
  (CBasic as Story).storyName = '0. 基础用法';
  (CBasic as Story).parameters = {
    docs: {
      description: {
        story: '简单的用法。\n\n',
      },
    },
  };
}
if (!(CButton as Story).storyName) {
  (CButton as Story).storyName = '10. 按钮下拉框';
  (CButton as Story).parameters = {
    docs: {
      description: {
        story: '使用`<Dropdown.Button>`可以使用具有下拉框功能的按钮。\n\n',
      },
    },
  };
}
if (!(CContextMenu as Story).storyName) {
  (CContextMenu as Story).storyName = '4. 右键触发菜单';
  (CContextMenu as Story).parameters = {
    docs: {
      description: {
        story: '点击鼠标右键后触发下拉菜单。\n\n',
      },
    },
  };
}
if (!(CEvent as Story).storyName) {
  (CEvent as Story).storyName = '5. 触发事件';
  (CEvent as Story).parameters = {
    docs: {
      description: {
        story: '点击菜单项后会触发事件，用户可以通过相应的菜单项`key`进行不同的操作。\n\n',
      },
    },
  };
}
if (!(CItem as Story).storyName) {
  (CItem as Story).storyName = '2. 其他元素';
  (CItem as Story).parameters = {
    docs: {
      description: {
        story: '分割线和不可用菜单项。\n\n',
      },
    },
  };
}
if (!(COverlayVisible as Story).storyName) {
  (COverlayVisible as Story).storyName = '8. 菜单隐藏方式';
  (COverlayVisible as Story).parameters = {
    docs: {
      description: {
        story: '默认是点击关闭菜单，可以关闭此功能。\n\n',
      },
    },
  };
}
if (!(CPlacement as Story).storyName) {
  (CPlacement as Story).storyName = '1. 弹出位置';
  (CPlacement as Story).parameters = {
    docs: {
      description: {
        story: '支持 12 个不同的方位。\n\n',
      },
    },
  };
}
if (!(CSubMenu as Story).storyName) {
  (CSubMenu as Story).storyName = '7. 多级菜单';
  (CSubMenu as Story).parameters = {
    docs: {
      description: {
        story: '传入的菜单里有多个层级。\n\n',
      },
    },
  };
}
if (!(CTrigger as Story).storyName) {
  (CTrigger as Story).storyName = '3. 点击触发菜单';
  (CTrigger as Story).parameters = {
    docs: {
      description: {
        story: '默认是移入触发下拉菜单，可以设置为点击触发。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/Dropdown/官网示例',
  tags: ['autodocs'],
};

export { CBasic, CButton, CContextMenu, CEvent, CItem, COverlayVisible, CPlacement, CSubMenu, CTrigger };
