// @universe-design/react-utilities/lib-commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "DomainEvent", {
  enumerable: true,
  get: function () {
    return _dom.DomainEvent;
  }
});
Object.defineProperty(exports, "DomainEventHandler", {
  enumerable: true,
  get: function () {
    return _react.DomainEventHandler;
  }
});
Object.defineProperty(exports, "DomainEventTarget", {
  enumerable: true,
  get: function () {
    return _dom.DomainEventTarget;
  }
});
Object.defineProperty(exports, "DomainEventTargetProvider", {
  enumerable: true,
  get: function () {
    return _react.DomainEventTargetProvider;
  }
});
Object.defineProperty(exports, "FindDOMNode", {
  enumerable: true,
  get: function () {
    return _useCloneElementRef.FindDOMNode;
  }
});
Object.defineProperty(exports, "KeyboardKey", {
  enumerable: true,
  get: function () {
    return _util.KeyboardKey;
  }
});
Object.defineProperty(exports, "ResizeObserver", {
  enumerable: true,
  get: function () {
    return _ResizeObserver.ResizeObserver;
  }
});
Object.defineProperty(exports, "StyleSheetOnce", {
  enumerable: true,
  get: function () {
    return _useStyleSheetOnce.StyleSheetOnce;
  }
});
Object.defineProperty(exports, "antdCompatibleDevWarning", {
  enumerable: true,
  get: function () {
    return _warning.compatibleDevWarning;
  }
});
Object.defineProperty(exports, "compatibleDevWarning", {
  enumerable: true,
  get: function () {
    return _warning.compatibleDevWarning;
  }
});
Object.defineProperty(exports, "composeRefs", {
  enumerable: true,
  get: function () {
    return _composeRefs.composeRefs;
  }
});
Object.defineProperty(exports, "createContextHook", {
  enumerable: true,
  get: function () {
    return _reactProvider.createContextHook;
  }
});
Object.defineProperty(exports, "createDescendantContext", {
  enumerable: true,
  get: function () {
    return _useDescendant.createDescendantContext;
  }
});
Object.defineProperty(exports, "createDescendantContextOptional", {
  enumerable: true,
  get: function () {
    return _useDescendant.createDescendantContextOptional;
  }
});
Object.defineProperty(exports, "createInOutTransition", {
  enumerable: true,
  get: function () {
    return _createInOutTransition.createInOutTransition;
  }
});
Object.defineProperty(exports, "createResizeObserver", {
  enumerable: true,
  get: function () {
    return _useResizeObserver.createResizeObserver;
  }
});
Object.defineProperty(exports, "defaultIsEqual", {
  enumerable: true,
  get: function () {
    return _useControlledState.defaultIsEqual;
  }
});
Object.defineProperty(exports, "deprecatedWarning", {
  enumerable: true,
  get: function () {
    return _warning.deprecatedWarning;
  }
});
Object.defineProperty(exports, "devWarning", {
  enumerable: true,
  get: function () {
    return _warning.devWarning;
  }
});
Object.defineProperty(exports, "forwardRef", {
  enumerable: true,
  get: function () {
    return _util.forwardRef;
  }
});
Object.defineProperty(exports, "getResizeObserver", {
  enumerable: true,
  get: function () {
    return _useResizeObserver.getResizeObserver;
  }
});
Object.defineProperty(exports, "getSize", {
  enumerable: true,
  get: function () {
    return _util.getSize;
  }
});
Object.defineProperty(exports, "getSizeAbbr", {
  enumerable: true,
  get: function () {
    return _util.getSizeAbbr;
  }
});
Object.defineProperty(exports, "isClient", {
  enumerable: true,
  get: function () {
    return _isClient.isClient;
  }
});
Object.defineProperty(exports, "isElement", {
  enumerable: true,
  get: function () {
    return _isElement.isElement;
  }
});
Object.defineProperty(exports, "isFunction", {
  enumerable: true,
  get: function () {
    return _util.isFunction;
  }
});
Object.defineProperty(exports, "isHTMLElement", {
  enumerable: true,
  get: function () {
    return _isHTMLElement.isHTMLElement;
  }
});
Object.defineProperty(exports, "isSVGElement", {
  enumerable: true,
  get: function () {
    return _isSVGElement.isSVGElement;
  }
});
Object.defineProperty(exports, "microDebounce", {
  enumerable: true,
  get: function () {
    return _util.microDebounce;
  }
});
Object.defineProperty(exports, "resetWarned", {
  enumerable: true,
  get: function () {
    return _warning.resetWarned;
  }
});
Object.defineProperty(exports, "resizeObservable", {
  enumerable: true,
  get: function () {
    return _resizeObservable.resizeObservable;
  }
});
Object.defineProperty(exports, "sleep", {
  enumerable: true,
  get: function () {
    return _sleep.sleep;
  }
});
Object.defineProperty(exports, "supportRef", {
  enumerable: true,
  get: function () {
    return _composeRefs.supportRef;
  }
});
Object.defineProperty(exports, "updateRef", {
  enumerable: true,
  get: function () {
    return _composeRefs.updateRef;
  }
});
Object.defineProperty(exports, "use", {
  enumerable: true,
  get: function () {
    return _use.use;
  }
});
Object.defineProperty(exports, "useCheckbox", {
  enumerable: true,
  get: function () {
    return _checkbox.useCheckbox;
  }
});
Object.defineProperty(exports, "useCloneElementRef", {
  enumerable: true,
  get: function () {
    return _useCloneElementRef.useCloneElementRef;
  }
});
Object.defineProperty(exports, "useConstant", {
  enumerable: true,
  get: function () {
    return _useConstant.useConstant;
  }
});
Object.defineProperty(exports, "useControlledState", {
  enumerable: true,
  get: function () {
    return _useControlledState.useControlledState;
  }
});
Object.defineProperty(exports, "useDomainEventTargetContext", {
  enumerable: true,
  get: function () {
    return _react.useDomainEventTargetContext;
  }
});
Object.defineProperty(exports, "useEffectOnce", {
  enumerable: true,
  get: function () {
    return _useEffectOnce.useEffectOnce;
  }
});
Object.defineProperty(exports, "useEventCallback", {
  enumerable: true,
  get: function () {
    return _useEventCallback.useEventCallback;
  }
});
Object.defineProperty(exports, "useForceUpdate", {
  enumerable: true,
  get: function () {
    return _useForceUpdate.useForceUpdate;
  }
});
Object.defineProperty(exports, "useIsomorphicLayoutEffect", {
  enumerable: true,
  get: function () {
    return _useIsomorphicLayoutEffect.useIsomorphicLayoutEffect;
  }
});
Object.defineProperty(exports, "useLatest", {
  enumerable: true,
  get: function () {
    return _useLatest.useLatest;
  }
});
Object.defineProperty(exports, "useLatestState", {
  enumerable: true,
  get: function () {
    return _useLatestState.useLatestState;
  }
});
Object.defineProperty(exports, "useLogger", {
  enumerable: true,
  get: function () {
    return _useLogger.useLogger;
  }
});
Object.defineProperty(exports, "useMergedRefs", {
  enumerable: true,
  get: function () {
    return _composeRefs.useMergedRefs;
  }
});
Object.defineProperty(exports, "usePrevious", {
  enumerable: true,
  get: function () {
    return _usePrevious.usePrevious;
  }
});
Object.defineProperty(exports, "useProxyRef", {
  enumerable: true,
  get: function () {
    return _useProxyRef.useProxyRef;
  }
});
Object.defineProperty(exports, "useResizeObserver", {
  enumerable: true,
  get: function () {
    return _useResizeObserver.useResizeObserver;
  }
});
Object.defineProperty(exports, "useStyleSheetOnce", {
  enumerable: true,
  get: function () {
    return _useStyleSheetOnce.useStyleSheetOnce;
  }
});
Object.defineProperty(exports, "useWillUnmount", {
  enumerable: true,
  get: function () {
    return _useWillUnmount.useWillUnmount;
  }
});
Object.defineProperty(exports, "warn", {
  enumerable: true,
  get: function () {
    return _warning.warn;
  }
});
Object.defineProperty(exports, "warning", {
  enumerable: true,
  get: function () {
    return _warning.warning;
  }
});
var _isClient = require("./dom/isClient");
var _isElement = require("./dom/isElement");
var _isHTMLElement = require("./dom/isHTMLElement");
var _isSVGElement = require("./dom/isSVGElement");
var _composeRefs = require("./hooks/composeRefs");
var _useEventCallback = require("./hooks/useEventCallback");
var _useIsomorphicLayoutEffect = require("./hooks/useIsomorphicLayoutEffect");
var _useLatest = require("./hooks/useLatest");
var _useEffectOnce = require("./hooks/useEffectOnce");
var _reactProvider = require("@universe-design/react-provider");
var _useProxyRef = require("./hooks/useProxyRef");
var _useLatestState = require("./hooks/useLatestState");
var _useResizeObserver = require("./hooks/useResizeObserver");
var _resizeObservable = require("./dom/resizeObservable");
var _useControlledState = require("./hooks/useControlledState");
var _usePrevious = require("./hooks/usePrevious");
var _useStyleSheetOnce = require("./hooks/useStyleSheetOnce");
var _useConstant = require("./hooks/useConstant");
var _useForceUpdate = require("./hooks/useForceUpdate");
var _useWillUnmount = require("./hooks/useWillUnmount");
var _useLogger = require("./hooks/useLogger");
var _use = require("./hooks/use");
var _useCloneElementRef = require("./hooks/useCloneElementRef");
var _dom = require("./domain-events/dom");
var _react = require("./domain-events/react");
var _warning = require("./warning");
var _sleep = require("./others/sleep");
var _util = require("./util");
var _createInOutTransition = require("./components/createInOutTransition");
var _checkbox = require("./components/checkbox");
var _ResizeObserver = require("./components/ResizeObserver");
var _useDescendant = require("./descendant/use-descendant");
