// @universe-design/react-utilities/lib-commonjs/warning.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.compatibleDevWarning = compatibleDevWarning;
exports.deprecatedWarning = deprecatedWarning;
exports.devWarning = devWarning;
exports.resetWarned = resetWarned;
exports.warn = warn;
exports.warning = warning;
let warned = {};

/**
 * 打印一个警告。这个方法不会去重，只是单纯打印一个警告。
 *
 * @public
 * @param valid - 是否是合法行为，false 则会打印
 * @param message - 警告消息
 * @param args - 其他参数
 * @returns
 */
function warn(valid, message) {
  if (process.env.NODE_ENV !== 'production') {
    if (valid) {
      return;
    }
    const text = `Warning: ${message}`;
    if (typeof console !== 'undefined') {
      for (var _len = arguments.length, args = new Array(_len > 2 ? _len - 2 : 0), _key = 2; _key < _len; _key++) {
        args[_key - 2] = arguments[_key];
      }
      // eslint-disable-next-line no-console
      console.error(text, ...args);
    }

    // Throwing an error and catching it immediately
    // to improve debugging
    // A consumer can use 'pause on caught exceptions'
    // https://github.com/facebook/react/issues/4216
    try {
      throw Error(text);
      // eslint-disable-next-line no-empty
    } catch (x) {}
  }
}

/**
 * 重置警告 map。重置后可以打印之前打印过的警告。
 *
 * @public
 */
function resetWarned() {
  warned = {};
}
function call(method, valid, message) {
  for (var _len2 = arguments.length, args = new Array(_len2 > 3 ? _len2 - 3 : 0), _key2 = 3; _key2 < _len2; _key2++) {
    args[_key2 - 3] = arguments[_key2];
  }
  if (!valid && !warned[message]) {
    method(false, message, ...args);
    warned[message] = true;
  }
}

/**
 * 打印一个一次性警告。这个函数命名为 warning 可以配合 `babel-plugin-dev-expression` 插件进行转码加上 NODE_ENV 限制，最终 production 代码中没有任何的 warning 调用。
 *
 * @public
 * @param valid - 是否是合法行为，false 则会打印
 * @param message - 警告消息
 * @param args - 其他参数
 */
function warning(valid, message) {
  for (var _len3 = arguments.length, args = new Array(_len3 > 2 ? _len3 - 2 : 0), _key3 = 2; _key3 < _len3; _key3++) {
    args[_key3 - 2] = arguments[_key3];
  }
  call(warn, valid, message, ...args);
}

/**
 * 格式化出一个 UD 开发警告
 *
 * @public
 * @param component - 组件名称
 * @param message - 警告信息
 * @returns
 *
 * @example
 *
 * ```typescript
 * warning(valid, devWarning('Component', 'Cannot use this prop.'))
 * ```
 */
function devWarning(component, message) {
  return `[ud: ${component}] ${message}`;
}

/**
 * 格式化出一个 AntD 兼容性开发警告
 *
 * @public
 * @param component - 组件名称
 * @param propNames - 属性名
 * @returns
 */
function compatibleDevWarning(component, propNames) {
  return `[ud: ${component}] ${propNames.map(o => `\`props.${o}\``).join(', ')} ${propNames.length > 1 ? 'are' : 'is'} not implemented, just for compatibility with other DS.`;
}

/**
 * 格式化出一个即将废弃警告
 *
 * @public
 * @param component - 组件名称
 * @param propName - 原属性名
 * @param replacePropName - 替换属性名
 * @returns
 */
function deprecatedWarning(component, propName, replacePropName) {
  return `[ud: ${component}] \`${propName}\` is deprecated.${replacePropName ? ` Please use ${(Array.isArray(replacePropName) ? replacePropName : [replacePropName]).map(o => `\`${o}\``).join(', ')} instead.` : ''}`;
}
