// @universe-design/rc-table/es/hooks/useScrollbarSize.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { useCallback, useState } from 'react';
import { isHTMLElement } from '../utils/styleUtil';
var computeScrollbarSize = function computeScrollbarSize(container) {
  var style = getComputedStyle(container);
  // 如果元素尺寸为 0， 此时认为滚动条尺寸为 0 （ 当 table 被 display:none时，会出现此情况 ）
  if (container.clientWidth === 0 && container.clientHeight === 0) {
    return {
      right: 0,
      bottom: 0
    };
  }
  var horizontalBorderWidth = Math.round(parseFloat(style.borderLeftWidth)) + Math.round(parseFloat(style.borderRightWidth));
  var verticalBorderWidth = Math.round(parseFloat(style.borderTopWidth)) + Math.round(parseFloat(style.borderBottomWidth));
  var rightScrollbarSize = Math.abs(container.offsetWidth - container.clientWidth - horizontalBorderWidth);
  var bottomScrollbarSize = Math.abs(container.offsetHeight - container.clientHeight - verticalBorderWidth);
  /**
   * 在缩放等情况下，尺寸获取会产生小数，最终运算产生误差
   * 我们先简单认为不会有 宽度小于等于 2px 的滚动条
   */
  var Minimal_Scrollbar_Size = 2;
  return {
    right: rightScrollbarSize > Minimal_Scrollbar_Size ? rightScrollbarSize : 0,
    bottom: bottomScrollbarSize > Minimal_Scrollbar_Size ? bottomScrollbarSize : 0
  };
};
/**
 * 分析一下内容高度变化会不会影响 scrollbar 的宽度
 * 1. 系统常驻滚动条，overflow: overlay，设置 webkit-scrollbar，scrollbarWidth 永远都是 0
 * 2. 系统常驻滚动条，overflow: overlay，不设置 webkit-scrollbar，scrollbarWidth 永远都是 0
 * 3. 系统常驻滚动条，overflow: scroll，设置 webkit-scrollbar，scrollbarWidth 永远都是 webkit-scrollbar-width 宽
 * 4. 系统常驻滚动条，overflow: scroll，不设置 webkit-scrollbar，scrollbarWidth 永远都是系统滚动条宽
 * 5. 系统滚动时显示滚动条，overflow: overlay，（不）设置 webkit-scrollbar，scrollbarWidth 永远都是 0
 * 6. 系统滚动时显示滚动条，overflow: scroll，设置 webkit-scrollbar，scrollbarWidth 不会变，是 webkit-scrollbar-width 宽
 * 7. 系统滚动时显示滚动条，overflow: scroll，不设置 webkit-scrollbar，scrollbarWidth 在 mac 上永远是 0，在 window 上永远是系统宽度。（window 不管隐藏还是不隐藏 scroll 的时候都会显示 track，mac 没问题，因为 scroll 和 auto 行为一致都是 0）
 *
 * TODO 7 <--> 4 的切换 mac 可能会有导致表头列分割线对不齐的问题: 此问题在 firefox 上存在，因为 firefox 不支持 webkit-scrollbar
 * 如果不考虑上述情况，就可以用一个 cache 值只计算一次
 * 注意：overlay 的时候 mac 无法进行动态的常驻和非常驻切换，所以相当于不会有 7 <--> 4 问题
 */
export default function useScrollbarColumnWidth(containerRef) {
  var _useState = useState({
      right: 0,
      bottom: 0
    }),
    _useState2 = _slicedToArray(_useState, 2),
    scrollbarSize = _useState2[0],
    setScrollbarSize = _useState2[1];
  var calculateScrollbarColumnWidth = useCallback(function () {
    var container = containerRef.current;
    if (!container || !isHTMLElement(container)) {
      return;
    }
    var realScrollbarSize = computeScrollbarSize(container);
    setScrollbarSize(realScrollbarSize);
  }, []);
  return {
    scrollbarSize: scrollbarSize,
    calculateScrollbarColumnWidth: calculateScrollbarColumnWidth
  };
}
