// @universe-design/react/es/checkbox/Checkbox.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React, { useRef, forwardRef, useImperativeHandle } from 'react';
import clsx from 'clsx';
import stylex from '@universe-design/stylex';
import { useCheckbox } from '@universe-design/react-utilities';
import { useClassPrefix, useCssScope } from '../config-provider/hooks';
import { CheckboxCheckedSvg, CheckboxIndeterminateSvg } from './_assets/svgs';
import { useCheckboxGroupContext } from './useCheckboxContext';
import { isFunction, omit } from '../util';
import { markAsDisabledableComp } from '../shared/markAsDisabledableComp';
import { cssTypographyBody0 } from '../stylex';
import './style';

// feat 精细的布局，所有样式基于 input dome 的真实状态，具有最高的规范性和一致性

const CheckboxComp = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    onChange,
    onCheckChange,
    onMouseEnter,
    onMouseLeave,
    className,
    children,
    value,
    disabled,
    name,
    error,
    ...otherProps
  } = props;
  const inputRef = useRef(null);
  const prefix = useClassPrefix();
  const cssScope = useCssScope('checkbox');
  const context = useCheckboxGroupContext();
  const checkboxProps = {
    ...otherProps,
    className: clsx(stylex(cssTypographyBody0), className),
    disabled,
    name,
    error,
    classPrefix: `${prefix}__checkbox`,
    onChange: function (checked, e) {
      const eventParams = {
        target: {
          type: 'checkbox',
          checked,
          value
        },
        stopPropagation() {
          e.stopPropagation();
        },
        preventDefault() {
          e.preventDefault();
        },
        nativeEvent: e.nativeEvent,
        reactEvent: e
      };
      if (context !== null && context !== void 0 && context.onChange) {
        context.onChange(eventParams);
      }
      if (isFunction(onChange)) {
        onChange(eventParams);
      }
      if (isFunction(onCheckChange)) {
        onCheckChange(checked);
      }
    }
  };
  if (context) {
    var _context$value;
    checkboxProps.disabled = disabled || context.disabled;
    checkboxProps.name = context.name;
    checkboxProps.error = error || context.error;
    checkboxProps.checked = Boolean((_context$value = context.value) === null || _context$value === void 0 ? void 0 : _context$value.includes(value));
  } else if ('checked' in props) {
    checkboxProps.checked = Boolean(props.checked);
  }
  const {
    checked,
    indeterminate,
    inputProps,
    descriptionProps,
    wrapperProps
  } = useCheckbox(checkboxProps, inputRef);
  useImperativeHandle(ref, () => inputRef.current);

  /**
   * 由于 background 有渐变动画效果，但 checked-svg 与 indeterminate-svg 的切换是瞬间执行的，所以在切换的瞬间会出现只有 background 没有对勾或横杠的闪黑现象
   * 修改方案是: 同时返回 checked-svg 与 indeterminate-svg，而不是让两者做切换。在 checked 状态下将 checked-svg 的透明度置为 1 而 indeterminate-svg 的透明度置为 0；indeterminate 状态下同理
   */
  return /*#__PURE__*/React.createElement("label", _extends({}, wrapperProps, {
    className: wrapperProps.className,
    onMouseEnter: onMouseEnter,
    onMouseLeave: onMouseLeave
  }), /*#__PURE__*/React.createElement("span", {
    className: cssScope()
  }, /*#__PURE__*/React.createElement("input", _extends({}, omit(otherProps, ['error', 'className', 'style', 'defaultChecked', 'indeterminate']), inputProps, {
    ref: inputRef
  })), /*#__PURE__*/React.createElement("span", {
    className: cssScope('__wallpaper')
  }, /*#__PURE__*/React.createElement(CheckboxCheckedSvg, {
    className: clsx(cssScope('__checked-svg'), checked && !indeterminate && cssScope('__checked-svg-shown'))
  }), /*#__PURE__*/React.createElement(CheckboxIndeterminateSvg, {
    className: clsx(cssScope('__indeterminate-svg'), indeterminate && cssScope('__indeterminate-svg-shown'))
  }))), children ? /*#__PURE__*/React.createElement("span", descriptionProps, children) : null);
});
export const Checkbox = /* #__PURE__*/markAsDisabledableComp(CheckboxComp);
Checkbox.__UD_CHECKBOX = true;
if (process.env.NODE_ENV !== "production") {
  Checkbox.displayName = 'Checkbox';
}
