// @universe-design/react/es/checkbox/Group.js
import clsx from 'clsx';
import React, { forwardRef } from 'react';
import { useControlledState } from '@universe-design/react-utilities';
import { useCssScope } from '../config-provider/hooks';
import { CheckboxGroupContextProvider } from './context';
import { Checkbox } from './Checkbox';
import { isFunction } from '../util';
import './style';
const CheckboxGroup = /*#__PURE__*/forwardRef((props, ref) => {
  var _props$value;
  const groupCssScope = useCssScope('checkbox-group');
  const [value, setValue] = useControlledState(props.defaultValue, (_props$value = props.value) !== null && _props$value !== void 0 ? _props$value : [], {
    isControlledOverride: 'value' in props
  });
  const triggerOnChange = data => {
    isFunction(props.onChange) && props.onChange(data);
    !('value' in props) && setValue(data);
  };
  const isChecked = data => value === null || value === void 0 ? void 0 : value.includes(data);
  return /*#__PURE__*/React.createElement(CheckboxGroupContextProvider, {
    value: {
      onChange: e => {
        const {
          value: changeValue
        } = e.target;
        // 在 group 模式下，如果没有提供 val，不会触发 onChange。
        if (changeValue === undefined) {
          return;
        }
        if (value) {
          const result = [...value];
          const index = value.indexOf(changeValue);
          if (index === -1) {
            result.push(changeValue);
          } else {
            result.splice(index, 1);
          }
          triggerOnChange(result);
        } else {
          triggerOnChange([changeValue]);
        }
      },
      value,
      name: props.name,
      disabled: props.disabled,
      error: props.error
    }
  }, (() => {
    const {
      className,
      options,
      disabled,
      children,
      style,
      error,
      onMouseEnter,
      onMouseLeave
    } = props;
    let renderCheckbox = children;
    if (options && options.length > 0) {
      renderCheckbox = options.map(option => {
        if (typeof option === 'string') {
          return /*#__PURE__*/React.createElement(Checkbox, {
            key: option,
            disabled: disabled,
            value: option,
            checked: isChecked(option),
            error: error
          }, option);
        }
        return /*#__PURE__*/React.createElement(Checkbox, {
          key: `checkobx-option-${option.value}`,
          disabled: option.disabled || disabled,
          value: option.value,
          checked: isChecked(option.value),
          style: option.style,
          className: option.className,
          error: option.error || error
        }, option.label);
      });
    }
    const classNames = clsx(groupCssScope(), className);
    return /*#__PURE__*/React.createElement("div", {
      className: classNames,
      style: style,
      onMouseEnter: onMouseEnter,
      onMouseLeave: onMouseLeave,
      ref: ref
    }, renderCheckbox);
  })());
});
export default /*#__PURE__*/React.memo(CheckboxGroup);
