// @universe-design/rc-table/es/FixedHolder/index.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import _toConsumableArray from "@babel/runtime/helpers/esm/toConsumableArray";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
var _excluded = ["className", "noData", "columns", "flattenColumns", "colWidths", "columCount", "stickyOffsets", "direction", "fixHeader", "stickyTopOffset", "stickyBottomOffset", "stickyClassName", "onScroll", "maxContentScroll", "children"];
import * as React from 'react';
import { useMemo } from 'react';
import classNames from 'classnames';
import { fillRef } from "@universe-design/rc-util/es/ref";
import ColGroup from '../ColGroup';
import { useTableContextSelector } from '../context/TableContext';
function useColumnWidth(colWidths, columCount) {
  return useMemo(function () {
    var cloneColumns = [];
    for (var i = 0; i < columCount; i += 1) {
      var val = colWidths[i];
      if (val !== undefined) {
        cloneColumns[i] = val;
      } else {
        return null;
      }
    }
    return cloneColumns;
  }, [colWidths.join('_'), columCount]);
}
var useIntersectionObserver = function useIntersectionObserver() {
  var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var _options$disabled = options.disabled,
    disabled = _options$disabled === void 0 ? false : _options$disabled;
  var stickyPinnedDetectorRef = React.useRef();
  var _React$useState = React.useState(false),
    _React$useState2 = _slicedToArray(_React$useState, 2),
    pinned = _React$useState2[0],
    setPinned = _React$useState2[1];
  React.useEffect(function () {
    if (disabled) {
      return;
    }
    if (!window || !window.IntersectionObserver) {
      return;
    }
    var stickyPinnedDetector = stickyPinnedDetectorRef.current;
    if (!stickyPinnedDetector) return;
    var observer = new IntersectionObserver(function (e) {
      if (e[0].intersectionRatio <= 0) {
        setPinned(true);
      } else {
        setPinned(false);
      }
    });
    observer.observe(stickyPinnedDetector);
    return function () {
      observer.disconnect();
    };
  }, [disabled]);
  return [stickyPinnedDetectorRef, pinned];
};
var FixedHolder = /*#__PURE__*/React.forwardRef(function (_ref, ref) {
  var _classNames;
  var className = _ref.className,
    noData = _ref.noData,
    columns = _ref.columns,
    flattenColumns = _ref.flattenColumns,
    colWidths = _ref.colWidths,
    columCount = _ref.columCount,
    stickyOffsets = _ref.stickyOffsets,
    direction = _ref.direction,
    fixHeader = _ref.fixHeader,
    stickyTopOffset = _ref.stickyTopOffset,
    stickyBottomOffset = _ref.stickyBottomOffset,
    stickyClassName = _ref.stickyClassName,
    onScroll = _ref.onScroll,
    maxContentScroll = _ref.maxContentScroll,
    children = _ref.children,
    props = _objectWithoutProperties(_ref, _excluded);
  var prefixCls = useTableContextSelector(function (ctx) {
    return ctx.prefixCls;
  });
  var isSticky = useTableContextSelector(function (ctx) {
    return ctx.isSticky;
  });
  var scrollbarColumnWidth = useTableContextSelector(function (ctx) {
    return ctx.scrollbarColumnWidth;
  });
  var hasStickyTop = stickyTopOffset !== undefined;
  var hasStickyBottom = stickyBottomOffset !== undefined;
  // Pass wheel to scroll event
  var scrollRef = React.useRef(null);
  var setScrollRef = React.useCallback(function (element) {
    fillRef(ref, element);
    fillRef(scrollRef, element);
  }, [ref]);
  React.useEffect(function () {
    var scrollEl = scrollRef.current;
    if (!scrollEl) return;
    function onWheel(e) {
      var currentTarget = e.currentTarget,
        deltaX = e.deltaX;
      if (deltaX) {
        onScroll({
          currentTarget: currentTarget,
          scrollLeft: currentTarget.scrollLeft + deltaX
        });
        e.preventDefault();
      }
    }
    scrollEl.addEventListener('wheel', onWheel);
    return function () {
      scrollEl.removeEventListener('wheel', onWheel);
    };
  }, [onScroll]);
  // Check if all flattenColumns has width
  var allFlattenColumnsWithWidth = React.useMemo(function () {
    return flattenColumns.every(function (column) {
      return column.width >= 0;
    });
  }, [flattenColumns]);
  // Add scrollbar column
  var lastColumn = flattenColumns[flattenColumns.length - 1];
  var ScrollBarColumn = React.useMemo(function () {
    return {
      fixed: lastColumn ? lastColumn.fixed : null,
      scrollbar: true,
      onHeaderCell: function onHeaderCell() {
        return {
          className: "".concat(prefixCls, "-cell-scrollbar")
        };
      }
    };
  }, [lastColumn, prefixCls]);
  var columnsWithScrollbar = useMemo(function () {
    return scrollbarColumnWidth ? [].concat(_toConsumableArray(columns), [ScrollBarColumn]) : columns;
  }, [scrollbarColumnWidth, columns, ScrollBarColumn]);
  var flattenColumnsWithScrollbar = useMemo(function () {
    return scrollbarColumnWidth ? [].concat(_toConsumableArray(flattenColumns), [ScrollBarColumn]) : flattenColumns;
  }, [scrollbarColumnWidth, flattenColumns, ScrollBarColumn]);
  // 当 stickyHeader pinned 的时候，添加一个特殊的类名
  var _useIntersectionObser = useIntersectionObserver({
      disabled: !isSticky
    }),
    _useIntersectionObser2 = _slicedToArray(_useIntersectionObser, 2),
    stickyPinnedDetectorTopRef = _useIntersectionObser2[0],
    pinnedTop = _useIntersectionObser2[1];
  var _useIntersectionObser3 = useIntersectionObserver({
      disabled: !isSticky
    }),
    _useIntersectionObser4 = _slicedToArray(_useIntersectionObser3, 2),
    stickyPinnedDetectorBottomRef = _useIntersectionObser4[0],
    pinnedBottom = _useIntersectionObser4[1];
  // Calculate the sticky offsets
  var headerStickyOffsets = useMemo(function () {
    var right = stickyOffsets.right,
      left = stickyOffsets.left;
    return _objectSpread(_objectSpread({}, stickyOffsets), {}, {
      left: direction === 'rtl' ? [].concat(_toConsumableArray(left.map(function (width) {
        return width + scrollbarColumnWidth;
      })), [0]) : left,
      right: direction === 'rtl' ? right : [].concat(_toConsumableArray(right.map(function (width) {
        return width + scrollbarColumnWidth;
      })), [0]),
      isSticky: isSticky
    });
  }, [stickyOffsets, direction, isSticky, scrollbarColumnWidth]);
  var mergedColumnWidth = useColumnWidth(colWidths, columCount);
  var stickyPinnedDetectorStyle = {
    height: 1,
    backgroundColor: 'transparent'
  };
  return /*#__PURE__*/React.createElement(React.Fragment, null, hasStickyTop && /*#__PURE__*/React.createElement("div", {
    ref: stickyPinnedDetectorTopRef,
    style: stickyPinnedDetectorStyle
  }), /*#__PURE__*/React.createElement("div", {
    style: _objectSpread(_objectSpread(_objectSpread({
      overflow: 'hidden'
    }, isSticky ? {
      top: stickyTopOffset,
      bottom: stickyBottomOffset
    } : {}), hasStickyTop && {
      marginTop: pinnedTop ? 0 : -1
    }), hasStickyBottom && {
      marginBottom: pinnedBottom ? 0 : -1
    }),
    ref: setScrollRef,
    className: classNames(className, (_classNames = {}, _defineProperty(_classNames, stickyClassName, !!stickyClassName), _defineProperty(_classNames, "".concat(prefixCls, "__sticky--pinned-top"), pinnedTop), _defineProperty(_classNames, "".concat(prefixCls, "__sticky--pinned-bottom"), pinnedBottom), _classNames))
  }, /*#__PURE__*/React.createElement("table", {
    style: {
      tableLayout: 'fixed',
      visibility: noData || mergedColumnWidth ? null : 'hidden'
    }
  }, (!noData || !maxContentScroll || allFlattenColumnsWithWidth) && /*#__PURE__*/React.createElement(ColGroup, {
    colWidths: mergedColumnWidth ? [].concat(_toConsumableArray(mergedColumnWidth), [scrollbarColumnWidth]) : [],
    columCount: columCount + 1,
    columns: flattenColumnsWithScrollbar
  }), children(_objectSpread(_objectSpread({}, props), {}, {
    stickyOffsets: headerStickyOffsets,
    columns: columnsWithScrollbar,
    flattenColumns: flattenColumnsWithScrollbar
  })))), hasStickyBottom && /*#__PURE__*/React.createElement("div", {
    ref: stickyPinnedDetectorBottomRef,
    style: stickyPinnedDetectorStyle
  }));
});
FixedHolder.displayName = 'FixedHolder';
export default FixedHolder;
