// @universe-design/react/lib/hooks/useScrollStatus.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useScrollStatus = useScrollStatus;
var _react = require("react");
var _useImmer = require("use-immer");
var _getThunkElement = require("../util/getThunkElement");
/**
 * 检测滚动容器的滚动状态
 */
function useScrollStatus() {
  let params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  const {
    disableHorizontal = false,
    disableVertical = false
  } = params;
  const [overflow, updateOverflow] = (0, _useImmer.useImmer)({
    t: false,
    b: false,
    l: false,
    r: false
  });
  const checkOverflow = (0, _react.useCallback)(target => {
    const targetEl = (0, _getThunkElement.getThunkElement)(target);
    if (!targetEl) {
      return;
    }
    const {
      scrollTop,
      scrollHeight,
      clientHeight,
      scrollLeft,
      scrollWidth,
      clientWidth
    } = targetEl;
    const targetInset = {
      top: 0,
      bottom: clientHeight,
      left: 0,
      right: clientWidth
    };
    const childInset = {
      top: -scrollTop,
      bottom: -scrollTop + scrollHeight,
      left: -scrollLeft,
      right: -scrollLeft + scrollWidth
    };
    updateOverflow(prevOverflow => ({
      t: disableVertical ? prevOverflow.t : targetInset.top > childInset.top,
      b: disableVertical ? prevOverflow.b : targetInset.bottom < childInset.bottom,
      l: disableHorizontal ? prevOverflow.l : targetInset.left > childInset.left,
      r: disableHorizontal ? prevOverflow.r : targetInset.right < childInset.right
    }));
  }, [disableHorizontal, disableVertical, updateOverflow]);
  return {
    checkOverflow,
    overflow
  };
}
