// @universe-design/react/lib/hooks/useSize.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getPositionAndSize = getPositionAndSize;
exports.useSize = useSize;
var _react = _interopRequireWildcard(require("react"));
var _util = require("../util");
var _useMountedState = require("./useMountedState");
var _useIsomorphicLayoutEffect = require("./useIsomorphicLayoutEffect");
var _reactUtilities = require("@universe-design/react-utilities");
const defaults = {
  rect: {
    width: 0,
    height: 0,
    top: 0,
    left: 0
  }
};
function getPositionAndSize(target, entry) {
  const size = (0, _reactUtilities.getSize)(target, entry);
  const {
    left,
    top,
    bottom,
    right,
    x,
    y
  } = target.getBoundingClientRect();
  return {
    left,
    top,
    bottom,
    right,
    x,
    y,
    width: size.width,
    height: size.height
  };
}
function useSize() {
  let {
    onChange,
    resizeObserverDisabled = false
  } = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  const [rect, setRect] = _react.default.useState(defaults.rect);
  const isMounted = (0, _useMountedState.useMountedState)();
  const handleResize = (0, _react.useCallback)(_size => {
    if (!isMounted()) {
      return;
    }
    setRect(_rect => {
      if (_size.width !== _rect.width || _size.height !== _rect.height) {
        return {
          ..._rect,
          width: _size.width,
          height: _size.height
        };
      }
      return _rect;
    });
  }, [isMounted]);
  const handleReLayout = (0, _react.useCallback)(_position => {
    if (!isMounted()) {
      return;
    }
    setRect(_rect => {
      if (_position.left !== _rect.left || _position.top !== _rect.top) {
        return {
          ..._rect,
          left: _position.left,
          top: _position.top
        };
      }
      return _rect;
    });
  }, [isMounted]);
  const ref = (0, _reactUtilities.useResizeObserver)((entry, observer) => {
    const {
      width,
      height,
      left,
      top
    } = getPositionAndSize(entry.target, entry);
    handleResize({
      width,
      height
    });
    handleReLayout({
      left,
      top
    });
    (0, _util.isFunction)(onChange) && onChange(entry, observer);
  }, {
    disabled: resizeObserverDisabled
  });
  const updateRect = (0, _react.useCallback)(() => {
    if (ref.current) {
      const {
        width,
        height,
        left,
        top
      } = getPositionAndSize(ref.current);
      handleResize({
        width,
        height
      });
      handleReLayout({
        left,
        top
      });
    }
  }, [handleReLayout, handleResize, ref]);
  (0, _useIsomorphicLayoutEffect.useIsomorphicLayoutEffect)(() => {
    // 1. 对 inline non-replaced element 进行一个简易的兜底
    if (ref.current) {
      const {
        width,
        height,
        left,
        top
      } = getPositionAndSize(ref.current);
      handleResize({
        width,
        height
      });
      handleReLayout({
        left,
        top
      });
    }
  }, [handleReLayout, handleResize, ref]);
  return [ref, (0, _react.useMemo)(() => ({
    ...rect,
    x: rect.left,
    y: rect.top,
    right: rect.width + rect.left,
    bottom: rect.height + rect.top
  }), [rect]), {
    handleResize,
    handleReLayout,
    updateRect
  }];
}
