// @byted-service/bytedtrace/lib/consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEBUG_FLAGS = exports.SAMPLED_FLAGS = exports.OPENTRACING_SPAN_TYPE = exports.ADDR_NETWORK_KEY = exports.POST_TRACE_ROOT_TAG = exports.MQ_LAG_KEY = exports.PRODUCE_TS_KEY = exports.MQ_PARTITION_KEY = exports.PRODUCER_DC_KEY = exports.PRODUCER_ADDR_KEY = exports.PRODUCER_CLUSTER_KEY = exports.PRODUCER_SERVICE_KEY = exports.CONSUMER_GROUP_KEY = exports.BATCH_SPAN_MAX_FAIL_DURATION_KEY = exports.BATCH_SPAN_MAX_SUCC_DURATION_KEY = exports.BATCH_SPAN_FAIL_CNT_KEY = exports.BATCH_SPAN_SUCC_CNT_KEY = exports.IS_BATCH_SPAN_KEY = exports.TO_REMOTE_ADDR = exports.FROM_REMOTE_ADDR = exports.TO_METHOD_KEY = exports.METHOD_KEY = exports.TO_DC_KEY = exports.TO_CLUSTER_KEY = exports.TO_SERVICE_KEY = exports.FROM_DC_KEY = exports.FROM_CLUSTER_KEY = exports.FROM_SERVICE_KEY = exports.TO_SERVICE_TYPE_KEY = exports.SERVICE_TYPE_KEY = exports.STATUS_CODE_KEY = exports.BUSINESS_STATUS_CODE_KEY = exports.IS_ERROR_KEY = exports.NAME_KEY = exports.ENV_KEY = exports.LANGUAGE_KEY = exports.DEPLOY_STAGE_KEY = exports.TCE_NODE_LEVEL_TAG = exports.TCE_PHYSICAL_CLUSTER_TAG = exports.FRAMEWOKER = exports.POD_NAME_KEY = exports.IPV6_KEY = exports.IPV4_KEY = exports.CLUSTER_KEY = exports.DC_KEY = exports.PSM_KEY = exports.TXN_LEN_OVER_FLOW = exports.GLOBAL_TRACE_VERSION = exports.HTTP_ROUTE = exports.HTTP_METHOD = void 0;
exports.DEFAULT_HTTPTIMEOUT = exports.DEFAULT_AGENT_PROVIDER_QUEUE_SIZE = exports.DEFAULT_QUEUE_SIZE = exports.TCP_ADDR = exports.UNIX_ADDR = exports.DEFAULT_CONNECTION_TYPE = exports.CONTENT_COMPRESSION = exports.HEADER_COMPRESSION = exports.USER_ID = exports.RESERVED = exports.DEFAULT_USE_DIC = exports.DEFAULT_BIG_ENDIAN = exports.DEFAULT_PROTO_TYPE = exports.LANGUAGE = exports.SDK_VERSION = exports.OPENTRACING_LOG_EVENT = exports.CLIENT_SPAN_TYPE = exports.SERVER_SPAN_TYPE = exports.FOLLOWS_FROM_REF = exports.CHILD_OF_REF = exports.INIT_SOCKET_SF_KEY = exports.DEFAULT_NET_TYPE = exports.CONNECT_WITHIN_TIME_INTERVAL_ERROR = exports.BYTELOGSETNAME = exports.BYTE_LOG_SET_NAME = exports.BYTE_LOG_VERSION_BYTE = exports.BYTE_LOG_MAGIC_NUMBER = exports.RECONNECT_TIME_INTERVAL = exports.NET_TYPES = exports.STORE_TYPE = exports.METER_TYPE = exports.RATE_COUNTER_TYPE = exports.TIMER_TYPE = exports.COUNTER_TYPE = exports.SPAN_KEY = exports.SPAN_IDKEY = exports.LOG_ID_KEY = exports.SDK_VERSION_TAG = exports.LOWER_BOUND_SAMPLER = exports.PROBABILITY_SAMPLER = exports.ADAPTIVE_SAMPLER = exports.POST_TRACE = exports.DEBUG_SAMPLING = exports.INHERITANCE = exports.ORIGIN_SPAN_ID_KEY = exports.SAMPLING_RATE_KEY = exports.DEGRADE_TAG = exports.DISABLE_SAMPLE = exports.LOWER_BOUND_FLAGS = exports.POST_TRACE_FLAGS = void 0;
exports.DEFAULT_MAX_BAG_SIZE = exports.IS_DEBUG = exports.IS_BENCHMARK = exports.TRACE_BAGGAGE_HEADER_NAME = exports.TRACER_CONTEXT_HEADER_NAME_W3C = exports.TRACE_CONTEXT_HEADER_NAME_JAEGER = exports.JAEGER_BAGGAGE_HEADER_PREFIX = exports.JAEGER_BAGGAGE_HEADER = exports.JAEGER_DEBUG_HEADER = exports.TRACE_DEBUG_HEADER = exports.HTTP_HEADER = exports.TEXT_MAP = exports.METRICS_COMPONENT = exports.METRICS_TENANT = exports.LogLevels = exports.PropagatorTypes = exports.ReferenceTypes = exports.DEFAULT_EVENT_FACTORY = exports.LOG_LEVEL_KEY = exports.LOG_EVENT_TYPE = exports.LINKER = exports.DEFAULT_ENABLE_SPAN_LOG = exports.COMMA = exports.EQUAL = exports.SpanModel = exports.MAX_INT_64 = exports.MAX_CREDIT_FOR_STAT_COUNTER = exports.MAX_HIT_PER_SECOND_FOR_STAT_COUNTER = exports.MAX_OPERATION_FOR_LIMITER = exports.DEFAULT_WINDOW_EXPIRED_TIME_MS = exports.DEFAULT_WINDOW_NUM = exports.DEFAULT_WINDOW_DURING_MS = exports.MAX_OPERATION_FOR_SAMPLER = void 0;
// tslint:disable-next-line: no-var-requires
const { version } = require('../package.json');
exports.HTTP_METHOD = '_http_method';
exports.HTTP_ROUTE = '_http_route';
exports.GLOBAL_TRACE_VERSION = 2;
exports.TXN_LEN_OVER_FLOW = '[bytedTrace] transaction data len overflow';
exports.PSM_KEY = '_psm';
exports.DC_KEY = '_dc';
exports.CLUSTER_KEY = '_cluster';
exports.IPV4_KEY = '_ipv4';
exports.IPV6_KEY = '_ipv6';
exports.POD_NAME_KEY = '_pod_name';
exports.FRAMEWOKER = '_framework';
exports.TCE_PHYSICAL_CLUSTER_TAG = '_tce_physical_cluster';
exports.TCE_NODE_LEVEL_TAG = '_tce_node_level';
exports.DEPLOY_STAGE_KEY = '_deploy_stage';
exports.LANGUAGE_KEY = '_language';
exports.ENV_KEY = '_env';
exports.NAME_KEY = '_name';
exports.IS_ERROR_KEY = '_is_error';
exports.BUSINESS_STATUS_CODE_KEY = '_biz_status_code';
exports.STATUS_CODE_KEY = '_status_code';
exports.SERVICE_TYPE_KEY = '_service_type';
exports.TO_SERVICE_TYPE_KEY = '_to_service_type';
exports.FROM_SERVICE_KEY = '_from_service';
exports.FROM_CLUSTER_KEY = '_from_cluster';
exports.FROM_DC_KEY = '_from_dc';
exports.TO_SERVICE_KEY = '_to_service';
exports.TO_CLUSTER_KEY = '_to_cluster';
exports.TO_DC_KEY = '_to_dc';
exports.METHOD_KEY = '_method';
exports.TO_METHOD_KEY = '_to_method';
exports.FROM_REMOTE_ADDR = '_from_addr';
exports.TO_REMOTE_ADDR = '_to_addr';
exports.IS_BATCH_SPAN_KEY = '_batch_span';
exports.BATCH_SPAN_SUCC_CNT_KEY = '_succ_cnt';
exports.BATCH_SPAN_FAIL_CNT_KEY = '_fail_cnt';
exports.BATCH_SPAN_MAX_SUCC_DURATION_KEY = '_max_succ_duration';
exports.BATCH_SPAN_MAX_FAIL_DURATION_KEY = '_max_fail_duration';
exports.CONSUMER_GROUP_KEY = '_consumer_group';
exports.PRODUCER_SERVICE_KEY = '_producer_service';
exports.PRODUCER_CLUSTER_KEY = '_producer_cluster';
exports.PRODUCER_ADDR_KEY = '_producer_addr';
exports.PRODUCER_DC_KEY = '_producer_dc';
exports.MQ_PARTITION_KEY = '_mq_partition';
exports.PRODUCE_TS_KEY = '_produce_ts';
exports.MQ_LAG_KEY = '_mq_lag';
exports.POST_TRACE_ROOT_TAG = '_posttrace_root';
exports.ADDR_NETWORK_KEY = '_network';
exports.OPENTRACING_SPAN_TYPE = 'ospan';
exports.SAMPLED_FLAGS = 1; // 0000 0001
exports.DEBUG_FLAGS = 2; // 0000 0010
exports.POST_TRACE_FLAGS = 4; // 0000 0100
exports.LOWER_BOUND_FLAGS = 8;
exports.DISABLE_SAMPLE = 254; // 1111 1110
// Default tag name
exports.DEGRADE_TAG = '_degrade';
exports.SAMPLING_RATE_KEY = '_sr';
exports.ORIGIN_SPAN_ID_KEY = '_origin_span_id';
// Sampling type, others see Sampler's sampler type
exports.INHERITANCE = 0;
exports.DEBUG_SAMPLING = 1;
exports.POST_TRACE = 2;
// Sampler type
exports.ADAPTIVE_SAMPLER = 10;
exports.PROBABILITY_SAMPLER = 11;
exports.LOWER_BOUND_SAMPLER = 12;
exports.SDK_VERSION_TAG = '_sdk_version';
exports.LOG_ID_KEY = 'K_LOGID';
exports.SPAN_IDKEY = 'K_SPANID';
exports.SPAN_KEY = '_SpanKey';
// Metrics types
exports.COUNTER_TYPE = 'counter';
exports.TIMER_TYPE = 'timer';
exports.RATE_COUNTER_TYPE = 'rateCounter';
exports.METER_TYPE = 'meter';
exports.STORE_TYPE = 'store';
exports.NET_TYPES = {
    unix: { path: '/opt/tmp/ttlogagent/bytelog.sock' },
    tcp: { host: '127.0.0.1', port: 8080 },
};
exports.RECONNECT_TIME_INTERVAL = 1000;
// ByteLog protocol header:
// --------------------------------------------------------------------------------------
// |  2 bytes       |   1 byte    | 4 bytes |  4 bytes  |  1 byte       |  x bytes      |
// --------------------------------------------------------------------------------------
// |  magic number  |   version   | size    |  reserved |  log set size |  log set name |
// --------------------------------------------------------------------------------------
// ByteLog protocol magic number in header.
exports.BYTE_LOG_MAGIC_NUMBER = 0x11;
// ByteLog protocol version byte in header.
exports.BYTE_LOG_VERSION_BYTE = 1;
// ByteLog log set name in header.
exports.BYTE_LOG_SET_NAME = '_bytelog.bytedtrace';
exports.BYTELOGSETNAME = 'BYTEDTRACE_BYTELOGSETNAME';
exports.CONNECT_WITHIN_TIME_INTERVAL_ERROR = 'connect time within time interval';
// export  const UN_RECOGNIZE_NET_TYPE_ERROR = 'UnRecognize netType type';
exports.DEFAULT_NET_TYPE = 'unix';
exports.INIT_SOCKET_SF_KEY = 'INIT_SOCKET_SF_KEY';
exports.CHILD_OF_REF = 0;
exports.FOLLOWS_FROM_REF = 1;
exports.SERVER_SPAN_TYPE = 'server';
exports.CLIENT_SPAN_TYPE = 'client';
exports.OPENTRACING_LOG_EVENT = 'olog';
// SDK Version
exports.SDK_VERSION = `v${version}`;
exports.LANGUAGE = 'js';
// Proto
// 采用pb序列化
exports.DEFAULT_PROTO_TYPE = 0;
// 序列化器
// DefaultSerializerType = constant.FlowSerializer
// 是否大端存储
exports.DEFAULT_BIG_ENDIAN = false;
// 是否使用字典
exports.DEFAULT_USE_DIC = true;
// 保留字段
exports.RESERVED = 0;
// UserId, is zero for sdk
exports.USER_ID = 0;
// Attention: sdk不支持压缩,所以不可修改压缩类型
// 私有协议头是否压缩
exports.HEADER_COMPRESSION = 0;
// 私有协议content是否压缩
exports.CONTENT_COMPRESSION = 0;
// Transport
// 默认的连接方式为tcp
exports.DEFAULT_CONNECTION_TYPE = 'unix';
exports.UNIX_ADDR = '/opt/tmp/ttlogagent/bytelog.sock';
exports.TCP_ADDR = '127.0.0.1:8080';
// Transaction传输channel大小
exports.DEFAULT_QUEUE_SIZE = 1000;
// AgentProvider传输channel大小
exports.DEFAULT_AGENT_PROVIDER_QUEUE_SIZE = 1024;
// Http client连接超时
exports.DEFAULT_HTTPTIMEOUT = 5 * 1000;
// Sampler
// Adaptive SingleSampler
exports.MAX_OPERATION_FOR_SAMPLER = 1000;
exports.DEFAULT_WINDOW_DURING_MS = 100;
exports.DEFAULT_WINDOW_NUM = 100;
// 窗口过期时间, 10分钟
exports.DEFAULT_WINDOW_EXPIRED_TIME_MS = 60 * 10 * 1e3;
// Tracer
exports.MAX_OPERATION_FOR_LIMITER = 1000;
// StateCounter
exports.MAX_HIT_PER_SECOND_FOR_STAT_COUNTER = 10;
exports.MAX_CREDIT_FOR_STAT_COUNTER = 20;
exports.MAX_INT_64 = '9223372036854775807';
var SpanModel;
(function (SpanModel) {
    SpanModel[SpanModel["BaseSpanModel"] = 0] = "BaseSpanModel";
    SpanModel[SpanModel["BatchSpanModel"] = 1] = "BatchSpanModel";
    SpanModel[SpanModel["ConsumerSpanModel"] = 2] = "ConsumerSpanModel";
})(SpanModel = exports.SpanModel || (exports.SpanModel = {}));
exports.EQUAL = '=';
exports.COMMA = ',';
exports.DEFAULT_ENABLE_SPAN_LOG = true;
exports.LINKER = ':';
// event level
exports.LOG_EVENT_TYPE = 'log';
exports.LOG_LEVEL_KEY = '_log_level';
exports.DEFAULT_EVENT_FACTORY = 0;
var ReferenceTypes;
(function (ReferenceTypes) {
    ReferenceTypes[ReferenceTypes["CHILD_OF_REF"] = 0] = "CHILD_OF_REF";
    ReferenceTypes[ReferenceTypes["FOLLOWS_FROM_REF"] = 1] = "FOLLOWS_FROM_REF";
})(ReferenceTypes = exports.ReferenceTypes || (exports.ReferenceTypes = {}));
var PropagatorTypes;
(function (PropagatorTypes) {
    /**
     * carrier 存储的 key-value 不需要转换字符串格式
     */
    PropagatorTypes["TEXT_MAP"] = "textMap";
    /**
     * carrier 是存储的 key-value 需要转换字符串格式
     * 例如 http url 中有中文被转码了
     */
    PropagatorTypes["HTTP_HEADER"] = "HTTPHeader";
})(PropagatorTypes = exports.PropagatorTypes || (exports.PropagatorTypes = {}));
var LogLevels;
(function (LogLevels) {
    LogLevels[LogLevels["LEVEL_NO_DEFINE"] = -1] = "LEVEL_NO_DEFINE";
    LogLevels[LogLevels["LEVEL_TRACE"] = 0] = "LEVEL_TRACE";
    LogLevels[LogLevels["LEVEL_DEBUG"] = 1] = "LEVEL_DEBUG";
    LogLevels[LogLevels["LEVEL_INFO"] = 2] = "LEVEL_INFO";
    LogLevels[LogLevels["LEVEL_NOTICE"] = 3] = "LEVEL_NOTICE";
    LogLevels[LogLevels["LEVEL_WARN"] = 4] = "LEVEL_WARN";
    LogLevels[LogLevels["LEVEL_ERROR"] = 5] = "LEVEL_ERROR";
    LogLevels[LogLevels["LEVEL_FATAL"] = 6] = "LEVEL_FATAL";
})(LogLevels = exports.LogLevels || (exports.LogLevels = {}));
exports.METRICS_TENANT = 'bytedtrace';
exports.METRICS_COMPONENT = 'sdk';
// propagator
exports.TEXT_MAP = 'textMap';
exports.HTTP_HEADER = 'HTTPHeader';
exports.TRACE_DEBUG_HEADER = 'x-tt-trace';
// trace 1.0
exports.JAEGER_DEBUG_HEADER = 'jaeger-debug-id';
exports.JAEGER_BAGGAGE_HEADER = 'jaeger-baggage';
exports.JAEGER_BAGGAGE_HEADER_PREFIX = 'bytedctx-';
exports.TRACE_CONTEXT_HEADER_NAME_JAEGER = 'byted-trace-id';
// Header for bytedtrace
exports.TRACER_CONTEXT_HEADER_NAME_W3C = 'traceparent';
exports.TRACE_BAGGAGE_HEADER_NAME = 'tracestate';
exports.IS_BENCHMARK = process.env.BYTEDTRACE_NODE_JS_BENCHMARK === 'true';
exports.IS_DEBUG = process.env.BYTEDTRACE_DEBUG_MODE === 'true';
exports.DEFAULT_MAX_BAG_SIZE = exports.IS_DEBUG ? 1 : 1048576;
