// @byted-service/bytedtrace/lib/transaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transaction = void 0;
const tslib_1 = require("tslib");
const consts = tslib_1.__importStar(require("./consts"));
const types = tslib_1.__importStar(require("./types"));
const status = tslib_1.__importStar(require("./status"));
const utils = tslib_1.__importStar(require("./utils"));
const bytetracePb = tslib_1.__importStar(require("./gen/bytedtrace_pb"));
const span_1 = require("./span");
const tracer_1 = require("./tracer");
const logger_1 = require("./logger");
const IS_FINISHED = 1;
class Transaction {
    constructor(tracer) {
        this.logId = '';
        this.transactionId = 1;
        this.method = '';
        this.samplingRate = 0;
        this.debugInfo = '';
        this.rootSpanId = '0';
        this.tags = {};
        this.pTraceType = 0;
        this.samplingType = 0;
        /**
         * 调用 setFinish 后，hasFinish 的值被设置为 1
         * 之后调用 isFinished 返回 true
         */
        this.hasFinish = 0;
        this.flags = 0;
        this.tracer = tracer;
        this.traceID = new tracer_1.TraceId('0', '0');
        this.reset();
    }
    setRootSpan(span) {
        this.rootSpan = span;
        this.rootSpanId = span.spanId;
    }
    setSamplingInfo(samplerType, samplingRate) {
        this.samplingRate = samplingRate;
        this.samplingType = samplerType;
    }
    isFinished() {
        return this.hasFinish === IS_FINISHED;
    }
    reset() {
        this.serverSpans = [];
        this.clientSpans = [];
        this.customSpans = [];
    }
    getSpanCount() {
        return (this.serverSpans.length +
            this.clientSpans.length +
            this.customSpans.length);
    }
    reportSpan(span) {
        span._reportMetricsAndLog();
        const { tracer } = this;
        const reported = this.shouldReported();
        const isRootSpan = span.getIsRootSpan();
        if (this.isFinished()) {
            if (reported) {
                const txn = this.copy();
                txn.doAppendSpan(span);
                tracer.options.reporter.report(txn);
            }
            return;
        }
        if (!reported) {
            if (isRootSpan) {
                this.setFinished();
            }
            return;
        }
        if (!this.appendSpan(span)) {
            return;
        }
        if (isRootSpan) {
            this.setFinished();
            tracer.options.reporter.report(this);
        }
    }
    appendSpan(span) {
        if (span.getIsRootSpan() ||
            this.getSpanCount() <
                this.tracer.options.breaker.getMaxSpanCountInTxn()) {
            this.doAppendSpan(span);
            return true;
        }
        this.tracer.options.statusCounter.recordWarningWithRateLimit(false, status.ErrAppendSpanNumOverflow, 'append span failed,exceed cap!');
        this.registerDegrade('append span exceed cap.');
        return false;
    }
    doAppendSpan(span) {
        if (span.isServerSpan()) {
            this.serverSpans.push(span);
        }
        else if (span.isClientSpan()) {
            this.clientSpans.push(span);
        }
        else {
            this.customSpans.push(span);
        }
    }
    setTagKv(key, value) {
        if (
        // TODO 这里应该需要加一个属性记录当前长度
        Object.keys(this.tags).length <
            this.tracer.options.breaker.getMaxSpanCountInTxn()) {
            this.tags[key] = value;
        }
        else {
            this.tracer.options.statusCounter.recordWarningWithRateLimit(false, status.ErrTxnTagOverflow, 'transaction tagKv number exceed cap!');
        }
    }
    setUniqueTagKv(key, value) {
        // TODO 这里应该有个 length 属性记录
        const len = Object.keys(this.tags).length;
        if (len > 0 && key in this.tags) {
            return;
        }
        if (len < this.tracer.options.breaker.getMaxTagKVCountInObj()) {
            this.tags[key] = value;
        }
        else {
            /* istanbul ignore next */
            this.tracer.options.statusCounter.recordWarningWithRateLimit(false, status.ErrTxnTagOverflow, 'transaction tagKv number exceed cap!');
        }
    }
    setBaggageItem(key, value) {
        this.baggage = this.baggage || new Map();
        this.baggage.set(key, value);
    }
    /**
     * 目前暂时只用到了 serializeBinary
     * @param protoType
     * @returns
     */
    toBs(protoType = types.ProtoType.pb) {
        const tpb = this.toPb();
        /* istanbul ignore if */
        if (consts.IS_DEBUG) {
            logger_1.debugLogger.debug('toBs tpb: %s', JSON.stringify(tpb.toObject()));
        }
        return tpb.serializeBinary();
    }
    genTraceID() {
        const traceID = new bytetracePb.TraceID();
        const high = this.traceID.high;
        const low = this.traceID.low;
        traceID.setHigh(high);
        traceID.setLow(low);
        return traceID;
    }
    initSampled(operationName, lowerBound) {
        if (utils.isSampledFlag(this.flags)) {
            return true;
        }
        if (this.passTxnSampling(operationName)) {
            this.flags = this.flags | consts.SAMPLED_FLAGS;
            if (lowerBound) {
                this.flags = this.flags | consts.LOWER_BOUND_FLAGS;
            }
            this.setBaggageItem(consts.SAMPLING_RATE_KEY, this.samplingRate + '');
            return true;
        }
        else {
            this.tracer.options.statusCounter.recordWarningWithRateLimit(false, status.ErrTxnSamplingOverflow, 'sampling transaction over limit,operation:%v', operationName);
        }
        return false;
    }
    initRootSpanContext(ctx, rate, operationName, needLimit) {
        let flags = ctx.flags;
        if (utils.isSampledFlag(flags) && needLimit) {
            if (!this.passTxnSampling(operationName)) {
                this.tracer.options.statusCounter.recordWarningWithRateLimit(false, status.ErrTxnSamplingOverflow, 'sampling transaction over limit,operation: %s', operationName);
                flags = utils.disableSampledFlag(flags);
            }
        }
        this.flags = flags;
        this.baggage = ctx.baggage;
        this.traceID = ctx.traceID;
        const samplingRate = ctx.getBaggageItem(consts.SAMPLING_RATE_KEY);
        const r = parseFloat(samplingRate);
        if (!Number.isNaN(r)) {
            this.setSamplingInfo(consts.INHERITANCE, Math.max(rate, r));
        }
        if (ctx.isDebug()) {
            this.debugInfo = ctx.debugInfo;
            this.enableDebugTrace(operationName, span_1.SamplingTraceType.debugTraceType, true);
        }
    }
    isBothSideTrace() {
        return this.pTraceType === span_1.SamplingTraceType.bothSideTraceType;
    }
    isDebugTrace() {
        if ((this.flags & consts.DEBUG_FLAGS) === consts.DEBUG_FLAGS) {
            return true;
        }
        return (this.pTraceType === span_1.SamplingTraceType.debugTraceType ||
            this.pTraceType === span_1.SamplingTraceType.bothSideTraceType);
    }
    enableDebugTrace(operationName, code, recordDebugTraceType) {
        if (this.tracer.isDyeDisable() || this.isDebugTrace()) {
            return;
        }
        const { breaker, statusCounter } = this.tracer.options;
        if (!breaker.debugTraceLimiterPack.pass(operationName)) {
            statusCounter.recordWarningWithRateLimit(false, status.ErrDebugTraceOverflow, 'enable DebugTrace over limit,operation:%s', this.method);
            return;
        }
        if (recordDebugTraceType && !this.passTxnSampling(operationName)) {
            statusCounter.recordWarningWithRateLimit(false, status.ErrTxnSamplingOverflow, 'sampling transaction over limit,operation:%s', operationName);
            return;
        }
        this.pTraceType = code;
        if (recordDebugTraceType) {
            this.samplingType = consts.DEBUG_SAMPLING;
            this.tracer._recordRootSampled(status.ROOT_DEBUG_QPS);
            if (!this.traceID || !this.traceID.isValid()) {
                this.traceID = this.tracer._genTraceId();
            }
        }
    }
    setDebugTrace(code, recordDebugTraceType) {
        this.enableDebugTrace(this.method, code, recordDebugTraceType);
    }
    setPostTrace(ctx) {
        if (!this.enablePostTrace(this.method, ctx)) {
            return false;
        }
        if (ctx &&
            (ctx.flags & consts.POST_TRACE_FLAGS) === consts.POST_TRACE_FLAGS &&
            (ctx.flags & consts.DEBUG_FLAGS) === consts.DEBUG_FLAGS) {
            this.enableDebugTrace(this.method, span_1.SamplingTraceType.bothSideTraceType, false);
        }
        else {
            this.pTraceType = span_1.SamplingTraceType.postTraceType;
        }
        return true;
    }
    enablePostTrace(operationName, ctx) {
        const { tracer, logId } = this;
        if (this.tracer.isPostTraceDisable() ||
            this.isPostTrace() ||
            this.isFinished()) {
            return false;
        }
        if (!this.passTxnSampling(operationName)) {
            this.tracer.options.statusCounter.recordWarningWithRateLimit(false, status.ErrTxnSamplingOverflow, 'sampling transaction over limit,operation:%v', this.method);
            return false;
        }
        // In w3c, there is no parent spanId in context, so parent spanId should be spanId-1
        this.rootSpan.setParentSpanId((BigInt(this.rootSpan.spanId) - BigInt(1)).toString(10));
        // PostTrace中traceID生成规则:
        // 如果上下文中有logID,根据logID生成traceID
        // 如果上下文中没有logID,选取下游传过来的context中的traceID
        // 如果下游没有传traceID,则随机生成一个TraceID
        if (logId && logId.length !== 0) {
            this.traceID = tracer.options.postTraceIDGenerator(logId);
        }
        else {
            if (ctx) {
                this.traceID = ctx.traceID;
            }
            else {
                this.traceID = this.tracer._genTraceId();
            }
        }
        if (!ctx) {
            this.tracer._recordRootSampled(status.ROOT_POST_TRACE_QPS);
        }
        this.samplingType = consts.POST_TRACE;
        return true;
    }
    setLocalSamplingTrace() {
        if (this.isSampled()) {
            return;
        }
        this.pTraceType = span_1.SamplingTraceType.localTraceType;
        if (!this.traceID.isValid()) {
            this.traceID = this.tracer._genTraceId();
        }
    }
    sampled() {
        return (this.flags & consts.SAMPLED_FLAGS) === consts.SAMPLED_FLAGS;
    }
    // 返回txn是否采样.
    // 注意:采样信息会传播.
    isSampled() {
        return this.sampled() || this.isSamplingPropagate();
    }
    // 返回采样是否传播,包括向上游和向下游传播.
    isSamplingPropagate() {
        const { pTraceType } = this;
        return (pTraceType === span_1.SamplingTraceType.postTraceType ||
            pTraceType === span_1.SamplingTraceType.debugTraceType ||
            pTraceType === span_1.SamplingTraceType.bothSideTraceType);
    }
    // 返回采样是否向下游传播.
    isForwardPropagate() {
        const flags = this.getFlags();
        if ((flags & consts.SAMPLED_FLAGS) !== consts.SAMPLED_FLAGS) {
            return false;
        }
        if ((flags & consts.DEBUG_FLAGS) === consts.DEBUG_FLAGS) {
            return true;
        }
        if ((flags & consts.POST_TRACE_FLAGS) !== consts.POST_TRACE_FLAGS) {
            return true;
        }
        return false;
    }
    isLocalSampled() {
        return this.pTraceType === span_1.SamplingTraceType.localTraceType;
    }
    isPostTrace() {
        return (this.pTraceType === span_1.SamplingTraceType.postTraceType ||
            this.pTraceType === span_1.SamplingTraceType.bothSideTraceType);
    }
    // 返回是否应该上报.
    // 上报和采样的区别:上报仅影响进程内的txn,采样会传播到上游或下游.
    // 以下三种情况下txn都会上报:
    // 1. 已经动态配置全量上报
    // 2. flags最末位为1(命中采样或者接受下游的postTrace)
    // 3. 已经被设置samplingTrace
    shouldReported() {
        if (this.tracer.options.disableReportSpan) {
            return false;
        }
        const allReported = this.tracer.allReported();
        const sampled = this.sampled();
        return allReported || sampled || this.pTraceType !== 0;
    }
    getFlags() {
        const ptrace = this.pTraceType;
        if (ptrace === 0) {
            return this.flags;
        }
        switch (ptrace) {
            case span_1.SamplingTraceType.localTraceType: {
                return this.flags | consts.DEBUG_FLAGS;
            }
            case span_1.SamplingTraceType.postTraceType: {
                return (this.flags | consts.POST_TRACE_FLAGS | consts.SAMPLED_FLAGS);
            }
            case span_1.SamplingTraceType.debugTraceType: {
                return this.flags | consts.DEBUG_FLAGS | consts.SAMPLED_FLAGS;
            }
            case span_1.SamplingTraceType.bothSideTraceType: {
                return (this.flags |
                    consts.POST_TRACE_FLAGS |
                    consts.DEBUG_FLAGS |
                    consts.SAMPLED_FLAGS);
            }
            default: /* istanbul ignore next */ {
                return this.flags;
            }
        }
    }
    toPb() {
        const txnPb = new bytetracePb.TransactionProto();
        txnPb.setTraceId(this.genTraceID());
        txnPb.setLogId(this.logId || '');
        txnPb.setTransactionId(this.transactionId);
        txnPb.setVersion(consts.GLOBAL_TRACE_VERSION);
        txnPb.setRootSpanId(this.rootSpanId);
        txnPb.setFlags(this.getFlags());
        const tags = [];
        const keys = Object.keys(this.tags);
        for (let i = 0; i < keys.length; i++) {
            const key = keys[i];
            const val = this.tags[key];
            tags.push([key, val]);
        }
        txnPb.setTagsList(this.tracer.options.breaker.buildTags(tags) || []);
        txnPb.setMethod(this.method);
        txnPb.setDebugInfo(this.debugInfo);
        txnPb.setSamplingRate(this.samplingRate);
        txnPb.setSamplingType(this.samplingType);
        if (this.serverSpans.length > 0) {
            const serverSpanPbs = this.serverSpans.map((s) => s._tranToServerSpanProto());
            txnPb.setServerSpansList(serverSpanPbs);
        }
        if (this.clientSpans.length > 0) {
            const clientSpanPbs = this.clientSpans.map((s) => s._tranToClientSpanProto());
            txnPb.setClientSpansList(clientSpanPbs);
        }
        if (this.customSpans.length > 0) {
            const customSpanPbs = this.customSpans.map((s) => s._tranToCustomSpanProto());
            txnPb.setCustomSpansList(customSpanPbs);
        }
        return txnPb;
    }
    passTxnSampling(operationName) {
        return this.tracer.options.breaker.txnSamplingLimiterPack.pass(operationName);
    }
    setFinished() {
        this.hasFinish = IS_FINISHED;
    }
    copy() {
        const txn = new Transaction(this.tracer);
        txn.transactionId = this.transactionId;
        txn.flags = this.flags;
        txn.pTraceType = this.pTraceType;
        txn.traceID = this.traceID;
        txn.method = this.method;
        txn.baggage = this.baggage;
        txn.debugInfo = this.debugInfo;
        txn.tags = this.tags;
        txn.rootSpanId = this.rootSpanId;
        txn.samplingType = this.samplingType;
        txn.samplingRate = this.samplingRate;
        return txn;
    }
    registerDegrade(str) {
        this.setUniqueTagKv(consts.DEGRADE_TAG, str);
    }
}
exports.Transaction = Transaction;
