// @byted-service/bytedtrace/lib/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AgentProviderStatus = exports.ReporterStatus = exports.ProtoType = exports.SpanType = void 0;
const consts_1 = require("./consts");
var SpanType;
(function (SpanType) {
    SpanType["server"] = "server";
    SpanType["client"] = "client";
    SpanType["custom"] = "custom";
})(SpanType = exports.SpanType || (exports.SpanType = {}));
var ProtoType;
(function (ProtoType) {
    ProtoType[ProtoType["pb"] = 0] = "pb";
    ProtoType[ProtoType["json"] = 50] = "json";
})(ProtoType = exports.ProtoType || (exports.ProtoType = {}));
var ReporterStatus;
(function (ReporterStatus) {
    ReporterStatus[ReporterStatus["running"] = 0] = "running";
    ReporterStatus[ReporterStatus["wait"] = 1] = "wait";
    ReporterStatus[ReporterStatus["close"] = 2] = "close";
})(ReporterStatus = exports.ReporterStatus || (exports.ReporterStatus = {}));
var AgentProviderStatus;
(function (AgentProviderStatus) {
    AgentProviderStatus[AgentProviderStatus["running"] = 0] = "running";
    AgentProviderStatus[AgentProviderStatus["wait"] = 1] = "wait";
    AgentProviderStatus[AgentProviderStatus["close"] = 2] = "close";
})(AgentProviderStatus = exports.AgentProviderStatus || (exports.AgentProviderStatus = {}));
