// @universe-design/react/lib/input/Group.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Group = void 0;
var React = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _util = require("../util");
var _hooks = require("../config-provider/hooks");
var _size = require("../config-provider/size");
require("./style");
/**
 * 组合输入框： `Input.Group`
 */

const Group = props => {
  const {
    style,
    className,
    size = 'medium',
    compact,
    children,
    onMouseEnter,
    onMouseLeave,
    onFocus,
    onBlur
  } = props;
  const cssScope = (0, _hooks.useCssScope)('input-group');
  const usedSize = (0, _util.getDefaultSizeAbbr)(size, 'medium');
  const cls = (0, _clsx.default)(cssScope(), {
    [cssScope(`--size-${usedSize}`)]: usedSize,
    [cssScope('--compact')]: compact
  }, className);
  return /*#__PURE__*/React.createElement(_size.InputGroupSizeContextProvider, {
    size: size
  }, /*#__PURE__*/React.createElement("span", {
    className: cls,
    style: style,
    onMouseEnter: onMouseEnter,
    onMouseLeave: onMouseLeave,
    onFocus: onFocus,
    onBlur: onBlur
  }, children));
};
exports.Group = Group;
var _default = exports.default = Group;
