// @universe-design/react/lib/input/Input.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Input = void 0;
var _u1n2z6e = _interopRequireDefault(require("./../.stylex/u1n2z6e7"));
var _u1u1p7mj = _interopRequireDefault(require("./../.stylex/u1u1p7mj"));
var _u184z4yl = _interopRequireDefault(require("./../.stylex/u184z4yl"));
var _u1981n0b = _interopRequireDefault(require("./../.stylex/u1981n0b"));
var _u8rtfz = _interopRequireDefault(require("./../.stylex/u8rtfz1"));
var _u9hwb1v = _interopRequireDefault(require("./../.stylex/u9hwb1v"));
var _u1xau0jj = _interopRequireDefault(require("./../.stylex/u1xau0jj"));
var _u5jxv = _interopRequireDefault(require("./../.stylex/u5jxv71"));
var _u99gbyh = _interopRequireDefault(require("./../.stylex/u99gbyh"));
var _urmav4p = _interopRequireDefault(require("./../.stylex/urmav4p"));
var _ukchsq = _interopRequireDefault(require("./../.stylex/ukchsq7"));
var _ubyhozq = _interopRequireDefault(require("./../.stylex/ubyhozq"));
var _u13d3rh = _interopRequireDefault(require("./../.stylex/u13d3rh8"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _EmptyFilled = _interopRequireDefault(require("@universe-design/icons-react/EmptyFilled"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _stylex2 = require("../stylex");
var _size = require("../util/size");
var _assert = require("../util/assert");
var _util = require("../util");
var _hooks = require("../config-provider/hooks");
var _Input = _interopRequireDefault(require("../shared/components/Input"));
var _Counter = _interopRequireDefault(require("./components/Counter"));
var _reactInput = require("@universe-design/react-input");
var _FadeTransition = require("../shared/components/transition/FadeTransition");
var _hooks2 = require("../hooks");
var _utils = require("./utils");
var _dom = require("../util/dom");
var _useTextOverflowPopperSidecar = require("../config-provider/hooks/useTextOverflowPopperSidecar");
var _react2 = require("@universe-design/stylex/react");
require("./style");
//.u11u7vat{min-width:20px;}
//.u1euv43f{position:absolute;}
//.u1xynx9j{left:0px;}
//.u113wtx2{top:0px;}
//.u5gq2j6{bottom:0px;}
//.u6dzj5z{max-width:100%;}
//.ud7fpy0{visibility:hidden;}
//.u10pi13n{position:relative;}
//.uqerorx{flex-grow:1;}
//.u10tiqix{min-width:0px;}
//.u22iagw{display:flex;}
//.u122n59{align-items:center;}
//.uly5x3f{width:100%;}
/** @jsxRuntime classic */
/** @jsx jsx */
/** @jsxFrag React.Fragment */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const Input = exports.Input = /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
  const {
    style,
    prefixCls,
    className,
    readOnly,
    placeholder,
    disabled,
    size,
    prefix,
    suffix,
    addonBefore,
    addonAfter,
    defaultValue = '',
    value: propsValue,
    allowClear,
    type,
    error,
    onChange,
    onKeyDown,
    onPressEnter,
    onClear,
    counter: counterProp,
    slicer,
    maxLength: maxLengthProp,
    showCount,
    truncateExcessWords,
    onCompositionStart: onCompositionStartProp,
    onCompositionEnd: onCompositionEndProp,
    sx,
    ...otherProps
  } = props;
  const cssScope = (0, _hooks.useCssScope)('input');
  const defaultSize = (0, _hooks.useComponentSize)();
  const inputGroupSize = (0, _hooks.useInputGroupSize)();
  const [focused, setFocused] = (0, _react.useState)(false);
  const inputRef = (0, _react.useRef)(null);
  const timeout = (0, _react.useRef)();
  const {
    isMouseOver,
    handleMouseEnter,
    handleMouseLeave
  } = (0, _hooks2.useMouseOver)();
  const currentSize = (0, _size.getDefaultSizeAbbr)(size || inputGroupSize || defaultSize);
  const tagSize = currentSize === 'lg' ? 'medium' : 'small';
  const handleInputChange = (0, _react.useCallback)((ref, event, value) => (0, _utils.resolveOnChange)(ref, event, onChange, value), [onChange]);
  const {
    nativeMaxLength,
    value,
    setValue,
    count,
    onCompositionEnd,
    onCompositionStart,
    internalOnChange
  } = (0, _reactInput.useInput)({
    maxLength: maxLengthProp,
    showCount,
    truncateExcessWords,
    counter: counterProp,
    slicer,
    defaultValue,
    inputRef,
    onChange: handleInputChange,
    onCompositionStart: onCompositionStartProp,
    onCompositionEnd: onCompositionEndProp,
    focused,
    ...('value' in props ? {
      value: propsValue
    } : {})
  });
  const placeholderRef = _react.default.useRef(null);
  const stop = (0, _react.useCallback)(() => {
    timeout.current && clearTimeout(timeout.current);
  }, []);
  const start = (0, _react.useCallback)(() => {
    timeout.current && clearTimeout(timeout.current);
    timeout.current = setTimeout(() => {
      if (inputRef.current && inputRef.current.getAttribute('type') === 'password' && inputRef.current.getAttribute('value')) {
        inputRef.current.removeAttribute('value');
      }
      // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    }, 13);
  }, []);
  (0, _react.useEffect)(() => {
    stop();
    start();
    return () => stop();
  }, [value, start, stop, type]);
  (0, _react.useImperativeHandle)(ref, () => inputRef.current);
  const renderAddonBefore = cls => {
    if (addonBefore) {
      return (0, _react2.jsx)("span", {
        className: `${cls}__addon-before`
      }, addonBefore);
    }
    return null;
  };
  const renderAddonAfter = cls => {
    if (addonAfter) {
      return (0, _react2.jsx)("span", {
        className: `${cls}__addon-after`
      }, addonAfter);
    }
    return null;
  };
  const renderPrefix = cls => {
    if (prefix) {
      return (0, _react2.jsx)("span", {
        className: `${cls}__prefix`
      }, prefix);
    }
    return null;
  };
  const renderSuffix = cls => {
    if (suffix || showCount) {
      return (0, _react2.jsx)("span", {
        className: `${cls}__suffix`
      }, suffix, (0, _react2.jsx)(_Counter.default
      // marginLeft 会根据 input Size 变化吗
      , {
        style: {
          marginLeft: suffix ? 8 : 0
        },
        showCount: showCount,
        count: count,
        maxLength: maxLengthProp,
        tagSize: tagSize
      }));
    }
    return null;
  };
  const onFocusChange = (0, _react.useCallback)(isFocus => {
    stop();
    start();
    setFocused(isFocus);
  }, [start, stop]);
  const handleClear = e => {
    var _inputRef$current;
    if (!('value' in props)) {
      setValue('');
    }
    if (inputRef.current) {
      (0, _utils.resolveOnChange)(inputRef.current, e, onChange, '');
    }
    (_inputRef$current = inputRef.current) === null || _inputRef$current === void 0 ? void 0 : _inputRef$current.focus();
    (0, _assert.isFunction)(onClear) && onClear();
  };
  const handleKeyDown = e => {
    /* istanbul ignore else  */
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    if (e.keyCode === 13 && onPressEnter) {
      onPressEnter(e);
    }
    /* istanbul ignore else  */
    if (onKeyDown) {
      onKeyDown(e);
    }
  };
  const renderClearIcon = cls => {
    if (allowClear && !disabled) {
      // 输入框激活态非 hover 状态不展示 query 清除按钮
      const isShow = isMouseOver && Boolean(value);
      const clx = (0, _clsx.default)({
        [`${cls}__clear-icon`]: true,
        [`${cls}__clear-icon-show`]: isShow
      });
      return (0, _react2.jsx)(_FadeTransition.FadeTransition, {
        in: isShow,
        timeout: 400
      }, (0, _react2.jsx)("span", {
        className: clx,
        onClick: handleClear
      }, (0, _react2.jsx)(_EmptyFilled.default, null)));
    }
    return null;
  };
  const TextOverflowPopper = (0, _useTextOverflowPopperSidecar.useTextOverflowPopperSidecar)();
  // 当value 为空时才应该显示 tooltip
  const isPlaceholderVisible = Boolean(!value && placeholder);
  const [isPlaceholderTooltipVisible, setIsPlaceholderTooltipVisible] = _react.default.useState(false);
  return (() => {
    const cls = prefixCls || cssScope();
    const rootCls = (0, _clsx.default)({
      [cls]: true,
      [`${cls}--error`]: error,
      [`${cls}-addon`]: addonAfter || addonBefore,
      [`${cls}--focused`]: focused,
      [`${cls}--disabled`]: disabled,
      [`${cls}--size-${currentSize}`]: currentSize
    }, className, (0, _stylex.default)({
      "props": {
        "a9b677": _u13d3rh.default
      }
    }, ...(sx !== null && sx !== void 0 ? sx : [])));
    return (0, _react2.jsx)("div", {
      style: style,
      className: rootCls
    }, renderAddonBefore(cls), (0, _react2.jsx)("label", {
      // 用 label 是为了嵌套 label 情况下能正常使用
      className: `${cls}-input-wrap`,
      onMouseEnter: handleMouseEnter,
      onMouseLeave: handleMouseLeave,
      onMouseDown: ev => {
        if (ev.target === inputRef.current) {
          // 如果是点击 input 的话就不需要做任何操作
          return;
        }
        // 如果事件是 wrap 里面的元素
        if ((0, _dom.eventContains)(ev.currentTarget, ev.nativeEvent)) {
          // 点击 wrap 内部元素不会让 input 失焦（mousedown 的 blur 逻辑会在这段同步代码之后执行）
          ev.preventDefault();
          // 点击 wrap 内部元素 focus 到 input 上
          (0, _util.focus)(inputRef, {
            selectTextIfInput: false
          });
        }
      },
      onClick: ev => {
        // 如果是 wrap 内部元素触发的 onClick，我们禁用 label 默认行为，因为 input 会触发 onClick，避免 bubble 两次（有对 prefix 和 suffix 误伤的概率）。
        if ((0, _dom.eventContains)(ev.currentTarget, ev.nativeEvent)) {
          ev.preventDefault();
        }
      }
    }, renderPrefix(cls), (0, _react2.jsx)(TextOverflowPopper, {
      title: (0, _react2.jsx)(_react.default.Fragment, null, placeholder),
      disabled: disabled,
      checkOverflow: () => {
        if (placeholderRef.current) {
          return (0, _dom.isOverflowed)(placeholderRef.current);
        }
        return false;
      },
      visible: isPlaceholderTooltipVisible && isPlaceholderVisible,
      onVisibleChange: visible => {
        setIsPlaceholderTooltipVisible(visible);
      }
    }, (0, _react2.jsx)("div", {
      className: (0, _clsx.default)(`${cls}-input__placeholder__wrapper`, (0, _stylex.default)({
        "props": {
          "qhf8xq": _u5jxv.default,
          "Bh6795r": _u99gbyh.default,
          "Bf4jedk": _urmav4p.default,
          "mc9l5x": _ukchsq.default,
          "Bt984gj": _ubyhozq.default
        }
      })),
      onMouseLeave: () => {
        /**
         * 在 mouseleave 时，可能不会触发 tooltip 的 visibleChange
         * 所以这里必须手动重置一下
         */
        setIsPlaceholderTooltipVisible(false);
      }
    }, (0, _react2.jsx)("div", {
      ref: placeholderRef,
      className: (0, _clsx.default)(`${cls}-input__placeholder`, (0, _stylex.default)(_stylex2.sxOverflowEllipsis, {
        "props": {
          "qhf8xq": _u1u1p7mj.default,
          "oyh7mz": _u184z4yl.default,
          "Bhzewxz": _u1981n0b.default,
          "B5kzvoi": _u8rtfz.default,
          "B2u0y6b": _u9hwb1v.default,
          "Bcdw1i0": _u1xau0jj.default
        }
      }))
    }, placeholder), (0, _react2.jsx)(_Input.default, (0, _extends2.default)({}, otherProps, {
      type: type,
      ref: inputRef,
      readOnly: readOnly,
      disabled: disabled,
      value: value,
      onChange: internalOnChange,
      onKeyDown: handleKeyDown,
      onFocusChange: onFocusChange,
      maxLength: nativeMaxLength,
      onCompositionEnd: onCompositionEnd,
      onCompositionStart: onCompositionStart,
      placeholder: placeholder,
      sx: [{
        "props": {
          "a9b677": _u13d3rh.default,
          "Bf4jedk": _u1n2z6e.default
        }
      }]
    })))), renderClearIcon(cls), renderSuffix(cls)), renderAddonAfter(cls));
  })();
});
if (process.env.NODE_ENV !== "production") {
  Input.displayName = 'Input';
}
