// @universe-design/react/lib/input/Password.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Password = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _Input = require("./Input");
var _VisibleOutlined = _interopRequireDefault(require("@universe-design/icons-react/VisibleOutlined"));
var _VisibleLockOutlined = _interopRequireDefault(require("@universe-design/icons-react/VisibleLockOutlined"));
var _hooks = require("../config-provider/hooks");
var _util = require("../util");
var _button = require("../button");
require("./style");
/**
 * 用于密码输入框 `Input.Password`
 */

const ActionMap = {
  click: 'onClick',
  hover: 'onMouseOver'
};
const Password = exports.Password = /*#__PURE__*/_react.default.forwardRef((props, ref) => {
  const {
    action = 'click',
    size,
    visibilityToggle = true,
    iconRender = visible => visible ? /*#__PURE__*/_react.default.createElement(_button.Button, {
      type: "icon",
      size: "small",
      icon: /*#__PURE__*/_react.default.createElement(_VisibleOutlined.default, null)
    }) : /*#__PURE__*/_react.default.createElement(_button.Button, {
      type: "icon",
      size: "small",
      icon: /*#__PURE__*/_react.default.createElement(_VisibleLockOutlined.default, null)
    }),
    ...restProps
  } = props;
  const [visible, setVisible] = (0, _react.useState)(false);
  const cssScope = (0, _hooks.useCssScope)('input-password');
  const prefixCls = cssScope();
  const onVisibleChange = () => {
    const {
      disabled
    } = props;
    if (disabled) {
      return;
    }
    setVisible(!visible);
  };
  const preventDefault = e => e.preventDefault();
  const suffixIcon = visibilityToggle && (() => {
    const iconTrigger = ActionMap[action];
    const icon = iconRender(visible);
    return /*#__PURE__*/_react.default.cloneElement(/*#__PURE__*/_react.default.isValidElement(icon) ? icon : /*#__PURE__*/_react.default.createElement("span", null, icon), {
      [iconTrigger]: onVisibleChange,
      key: 'passwordIcon',
      onMouseDown: e => preventDefault(e),
      onMouseUp: e => preventDefault(e)
    });
  })();
  const omittedProps = {
    ...(0, _util.omit)(restProps, ['suffix']),
    type: visible ? 'text' : 'password',
    prefixCls,
    suffix: suffixIcon
  };
  return /*#__PURE__*/_react.default.createElement(_Input.Input, (0, _extends2.default)({
    ref: ref
  }, omittedProps, {
    size: size
  }));
});
