// @universe-design/react/lib/input/Search.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Search = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _Input = require("./Input");
var _hooks = require("../config-provider/hooks");
var _SearchOutlineOutlined = _interopRequireDefault(require("@universe-design/icons-react/SearchOutlineOutlined"));
var _util = require("../util");
require("./style");
const SearchShapesEnum = ['round', 'square'];

/**
 * 用于搜索输入框 `Input.Search`
 * @public
 */

const Search = exports.Search = /*#__PURE__*/_react.default.forwardRef((props, ref) => {
  const {
    global: {
      search
    }
  } = (0, _hooks.useLocale)();
  const {
    shape = 'round',
    size: customerSize,
    placeholder = search,
    className,
    ...restProps
  } = props;
  const cssScope = (0, _hooks.useCssScope)('search');
  const configSize = (0, _hooks.useComponentSize)();
  const inputGroupSize = (0, _hooks.useInputGroupSize)();
  const size = customerSize || inputGroupSize || configSize;
  const usedSize = (0, _util.getDefaultSizeAbbr)(size);
  const cls = (0, _clsx.default)(cssScope(), {
    [cssScope(`--size-${usedSize}`)]: usedSize,
    [cssScope(`--shape-${shape}`)]: shape
  }, className);
  return /*#__PURE__*/_react.default.createElement(_Input.Input, (0, _extends2.default)({
    ref: ref,
    className: cls,
    size: size,
    placeholder: placeholder,
    prefix: /*#__PURE__*/_react.default.createElement(_SearchOutlineOutlined.default, null)
  }, restProps));
});
if (process.env.NODE_ENV !== "production") {
  Search.displayName = 'Search';
}
