// @universe-design/react/lib/input/TextArea.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TextArea = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _stylex2 = require("../stylex");
var _EmptyFilled = _interopRequireDefault(require("@universe-design/icons-react/EmptyFilled"));
var _hooks = require("../hooks");
var _reactUtilities = require("@universe-design/react-utilities");
var _hooks2 = require("../config-provider/hooks");
var _util = require("../util");
var _Counter = _interopRequireDefault(require("./components/Counter"));
var _reactInput = require("@universe-design/react-input");
require("./style");
var _utils = require("./utils");
/**
 * @public
 */

const TextArea = exports.TextArea = /*#__PURE__*/_react.default.forwardRef((props, ref) => {
  const {
    allowClear,
    autoSize,
    showCount,
    maxLength: maxLengthProp,
    value: valueProp,
    defaultValue,
    style,
    error,
    className,
    onKeyDown,
    onPressEnter,
    onChange,
    onClear,
    suffix,
    truncateExcessWords = false,
    counter: counterProp,
    slicer,
    onCompositionEnd: onCompositionEndProp,
    onCompositionStart: onCompositionStartProp,
    autoFocus,
    resize,
    ...resetProps
  } = props;
  const rootPrefixClass = (0, _hooks2.useClassPrefix)();
  const cssScope = (0, _hooks2.useCssScope)('textarea');
  const [textareaResized, setTextareaResized] = (0, _react.useState)(false);
  const textareaRef = (0, _reactUtilities.useResizeObserver)(entry => {
    /**
     * 监控 textarea 的 style 属性上是否有 height 属性
     * textarea 只有在拖拽改变大小之后才会有 style height 属性
     * 此时，更改父容器为 height: auto 以自适应高度
     */
    if (entry.target.style.height) {
      setTextareaResized(true);
    }
  }, {
    disabled: Boolean(autoSize)
  });
  const [textAreaStyle, setTextAreaStyle] = (0, _react.useState)({});
  const [focused, setFocused] = (0, _react.useState)(false);
  const handleInputChange = (0, _react.useCallback)((ref, event, value) => (0, _utils.resolveOnChange)(ref, event, onChange, value), [onChange]);
  const {
    nativeMaxLength,
    value,
    setValue,
    count,
    onCompositionEnd,
    onCompositionStart,
    internalOnChange,
    maxLength
  } = (0, _reactInput.useInput)({
    maxLength: maxLengthProp,
    showCount,
    truncateExcessWords,
    counter: counterProp,
    slicer,
    defaultValue,
    inputRef: textareaRef,
    onChange: handleInputChange,
    onCompositionStart: onCompositionStartProp,
    onCompositionEnd: onCompositionEndProp,
    focused,
    ...('value' in props ? {
      value: valueProp
    } : {})
  });
  (0, _reactInput.useFocus)({
    inputRef: textareaRef,
    onFocusChange: f => setFocused(f),
    autoFocus: Boolean(autoFocus)
  });
  (0, _react.useImperativeHandle)(ref, () => textareaRef.current);
  const adjustSize = (0, _react.useCallback)(() => {
    if (autoSize && textareaRef.current) {
      const config = {
        minRows: undefined,
        maxRows: undefined
      };
      if (typeof autoSize === 'object') {
        config.minRows = autoSize.minRows;
        config.maxRows = autoSize.maxRows;
      }
      const currentStyle = (0, _reactInput.calculateTextAreaHeight)(textareaRef.current, false, config.minRows, config.maxRows);
      if (currentStyle) {
        setTextAreaStyle(currentStyle);
      }
    } else {
      setTextAreaStyle({});
    }
  }, [autoSize, textareaRef]);
  const resizeRef = (0, _reactUtilities.useResizeObserver)(adjustSize); // , {disabled: !autoSize})

  (0, _hooks.useIsomorphicLayoutEffect)(adjustSize, [adjustSize, value]);
  const handleClear = e => {
    var _textareaRef$current;
    /* istanbul ignore else  */
    if (!('value' in props)) {
      setValue('');
    }
    /* istanbul ignore else  */
    if ((0, _util.isFunction)(onChange) && textareaRef.current) {
      const eventObj = Object.create(e);
      eventObj.target = textareaRef.current;
      eventObj.currentTarget = textareaRef.current;
      const originalValue = textareaRef.current.value;
      textareaRef.current.value = '';
      onChange(eventObj);
      textareaRef.current.value = originalValue;
    }
    (_textareaRef$current = textareaRef.current) === null || _textareaRef$current === void 0 ? void 0 : _textareaRef$current.focus();
    (0, _util.isFunction)(onClear) && onClear();
  };
  const showClearIcon = allowClear && !props.disabled && value;
  /**
   * 用户传入的 error true 优先级很高（可能存在输入值不符合某个 form 规则而不只是 maxLength 的情况）
   * count > maxLength error 的优先级也很高
   * 只要有其一就需要标红
   */
  const isError = error || (0, _util.isNonNullable)(maxLength) && count > maxLength;
  const cls = (0, _clsx.default)({
    [cssScope()]: true,
    [cssScope('--error')]: isError,
    [cssScope('--show-count')]: showCount,
    [cssScope('--show-clear')]: showClearIcon,
    [cssScope('--disabled')]: props.disabled
  }, className);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: cls,
    style: {
      ...style,
      ...(textareaResized ? {
        height: 'auto'
      } : {})
    },
    ref: resizeRef
  }, (() => {
    if (showClearIcon) {
      return /*#__PURE__*/_react.default.createElement("span", {
        className: cssScope('__clear-icon'),
        onClick: handleClear
      }, /*#__PURE__*/_react.default.createElement(_EmptyFilled.default, null));
    }
    return null;
  })(), /*#__PURE__*/_react.default.createElement("textarea", (0, _extends2.default)({}, resetProps, {
    style: {
      ...textAreaStyle,
      resize
    },
    ref: textareaRef,
    className: (0, _clsx.default)(`${rootPrefixClass}-scrollbar`, (0, _stylex.default)(_stylex2.cssTypographyBody0)),
    onChange: internalOnChange,
    onKeyDown: e => {
      /* istanbul ignore else  */
      if (e.keyCode === _util.KeyCode.ENTER) {
        (0, _util.isFunction)(onPressEnter) && onPressEnter(e);
      }
      (0, _util.isFunction)(onKeyDown) && onKeyDown(e);
    },
    onCompositionEnd: onCompositionEnd,
    onCompositionStart: onCompositionStart,
    value: value,
    maxLength: nativeMaxLength
  })), (() => /*#__PURE__*/_react.default.createElement("span", {
    className: cssScope('-suffix')
  }, suffix && /*#__PURE__*/_react.default.createElement("span", {
    className: cssScope('-suffix-slot')
  }, suffix), /*#__PURE__*/_react.default.createElement(_Counter.default, {
    showCount: Boolean(showCount),
    count: count,
    disabled: props.disabled,
    maxLength: maxLengthProp
  })))());
});
