// @universe-design/react/src/shared/components/picker/Context.ts
import type { Moment } from 'moment';
import type { Dispatch, ReactNode, SetStateAction } from 'react';
import type { Context } from 'use-context-selector';
import { createContext, useContext, useContextSelector } from 'use-context-selector';
import { devWarning, warning } from '../../../util';
import type { MODE } from './interface';

export interface PickerContextType<DateType> {
  mode: MODE;
  setMode: Dispatch<SetStateAction<MODE>>;
  value?: DateType;
  setValue: Dispatch<SetStateAction<DateType | undefined>>;
  setOpened: Dispatch<SetStateAction<boolean>>;
  showToday?: boolean;
  inDatePicker?: boolean;
  renderExtraFooter?: (ctx: Pick<PickerContextType<DateType>, 'mode' | 'value' | 'setValue'>) => ReactNode;
}

export const PickerContext = createContext<PickerContextType<any>>({
  mode: 'date',
  setMode: () => warning(false, devWarning('DatePicker', 'You are using the default setMode from context')),
  setValue: () => warning(false, devWarning('DatePicker', 'You are using the default setValue from context')),
  setOpened: () => warning(false, devWarning('DatePicker', 'You are using the default setOpened from context')),
  showToday: true,
});

export const getPickerContext = () => PickerContext;

export function usePickerContext<DateType = Moment>() {
  return useContext<PickerContextType<DateType>>(PickerContext as unknown as Context<PickerContextType<DateType>>);
}

export const usePickerContextSelector = <Selected, DateType = Moment>(
  selector: (value: PickerContextType<DateType>) => Selected,
): Selected => useContextSelector(PickerContext as unknown as Context<PickerContextType<DateType>>, selector);
