// @universe-design/react/src/shared/components/picker/interface.ts
import type { Dispatch, SetStateAction, ReactNode, CSSProperties } from 'react';
import type React from 'react';
import type { TriggerProps } from '@universe-design/rc-trigger';
import type { PickerLocale } from '../../locales';
import type { DefaultSizeType } from '../../../util';
// import { PickerContextType } from './Context';
import type { GenerateConfig } from './generate';
import type moment from 'moment';

export type MODE = 'date' | 'month' | 'year' | 'time' | 'timeCombine';

export type PickerType = 'date' | 'month' | 'year' | 'time';

export type Locale = PickerLocale;

export type IFormat<DateType> = string | ((value: DateType) => string);

export type IParse<DateType> = (value: string) => DateType;

export interface IFormatOption<DateType> {
  format?: IFormat<DateType>;
  mode?: MODE;
  locale?: Locale;
  localeName: string;
}

export type KeyboardConfig = {
  onLeftRight?: ((diff: number) => void) | null;
  onUpDown?: ((diff: number) => void) | null;
  onEnter?: (() => void) | null;
};

export type NullableDateType<DateType> = DateType | null | undefined;

export interface IParseOption<DateType> {
  parse?: IParse<DateType>;
  format?: IFormat<DateType>;
  mode?: MODE;
  locale?: Locale;
  localeName: string;
}

export type PickerRefConfig = {
  focus: () => void;
  blur: () => void;
};

export interface PickerState<DateType> {
  text?: string; // 选中后计算出的文本
  hoverText?: string; // 用户 hover 后计算出的文本
  value?: DateType;
  open?: boolean;
}

export type PickerTriggerElement<DateType> = ReactNode | ((state: PickerState<DateType>) => ReactNode);

export type PanelRefProps = {
  onKeyDown?: (e: React.KeyboardEvent<HTMLElement>) => boolean;
  onBlur?: React.FocusEventHandler<HTMLElement>;
  onClose?: () => void;
};
/**
 * @public
 */
export type CustomTimeType<DateType> = {
  /**
   * 值
   */
  value: DateType;
  /**
   * 显示的名称
   */
  label: string;
  /**
   * 是否禁用
   */
  disabled: boolean;
};

/** *
 * SharedXXX for internal & user
 * BaseXXX for internal only
 * Common for internal only and be reused
 */

export type CommonProps<DateType> = {
  value?: DateType;
  defaultValue?: DateType;
  mode?: MODE;
  format?: string;
  parse?: IParse<DateType>;
};

// -----------------------------Panel-----------------------------
export type SharedPanelProps<DateType> = {
  locale?: Locale;
  onOpenChange?: (open: boolean) => void;

  // event
  onPanelChange?: (value: DateType) => void;
  onKeyDown?: React.KeyboardEventHandler<HTMLInputElement>;
  onMouseDown?: React.MouseEventHandler<HTMLDivElement>;
  onMouseUp?: React.MouseEventHandler<HTMLDivElement>;
};

export type BasePanelProps<DateType> = {
  prefixCls: string;
  generateConfig: GenerateConfig<DateType>;
  picker: PickerType;
};

export type CommonPickerPanelProps<DateType> = {
  locale: Locale;
  onPanelChange: (value: DateType) => void;
} & CommonProps<DateType> &
  SharedPanelProps<DateType> &
  BasePanelProps<DateType>;

// -----------------------------Trigger-----------------------------
export type SharedTriggerProps<DateType> = {
  style?: CSSProperties;
  className?: string;
  size?: DefaultSizeType;
  error?: boolean;
  inputReadOnly?: boolean;
  suffixIcon?: React.ReactNode;
  autoFocus?: boolean;
  allowClear?: boolean;
  triggerElement?: PickerTriggerElement<DateType>;
  placeholder?: string;
  open?: boolean;
  popupStyle?: CSSProperties;
  getPopupContainer?: TriggerProps['getPopupContainer'];
  dropdownClassName?: string;
  disabled?: boolean;

  // event
  onClear?: () => void;
  onChange?: (value: DateType) => void;
  onClick?: React.MouseEventHandler<HTMLDivElement>;
};

export type BaseTriggerProps = {
  children: React.ReactElement;
  setOpened: Dispatch<SetStateAction<boolean>>;
  popupElement: React.ReactElement;
};

export type PickerTriggerProps<DateType> = SharedTriggerProps<DateType> & BaseTriggerProps & CommonProps<DateType>;

// -----------------------------TimePicker-----------------------------
export interface SharedTimeProps<DateType = moment.Moment> {
  /**
   * 小时选项间隔
   */
  hourStep?: number;
  /**
   * 分钟选项间隔
   */
  minuteStep?: number;
  /**
   * 秒选项间隔
   */
  secondStep?: number;
  /**
   * 禁用的小时选项
   */
  disabledHours?: () => number[];
  /**
   * 禁用的分钟选项
   */
  disabledMinutes?: (hour: number) => number[];
  /**
   * 禁用的秒选项
   */
  disabledSeconds?: (hour: number, minute: number) => number[];
  timeStep?: number;
  disabledTime?: string[];
  customTime?: CustomTimeType<DateType>[];
  /**
   * 是否隐藏禁用的选项
   */
  hideDisabledOptions?: boolean;
  /**
   * 是否显示小时
   */
  showHour?: boolean;
  /**
   * 是否显示分钟
   */
  showMinute?: boolean;
  /**
   * 是否显示秒
   */
  showSecond?: boolean;
  /**
   * 是否使用12小时时间制
   */
  use12Hours?: boolean;
  /**
   * 默认值
   */
  defaultValue?: DateType;
  /**
   * 使用的时间格式
   */
  format?: string;
}

/**
 * @docgen
 * @public
 */
export type TimePickerProps<DateType> = {
  picker: 'time';
} & PickerProps<DateType> &
  SharedTimeProps<DateType>;

export type CommonTimePanelProps<DateType> = {
  showHour?: boolean;
  showMinute?: boolean;
  showSecond?: boolean;
  hideDisabledOptions?: boolean;
  customTime?: CustomTimeType<DateType>[];
  use12Hours?: boolean;
  operationRef?: React.MutableRefObject<PanelRefProps>;
  activeColumnIndex?: number;
  popupStyle?: CSSProperties;
} & CommonPickerPanelProps<DateType>;

export type NormalTimePanelProps<DateType> = Omit<SharedTimeProps<DateType>, 'timeStep' | 'disabledTime'> &
  CommonTimePanelProps<DateType>;

export type CombineTimePanelProps<DateType> = {
  timeStep?: number;
  option?: IParseOption<DateType>;
  disabledTime?: string[];
} & CommonTimePanelProps<DateType>;

// 时间面板的所有props
export type TimePanelProps<DateType> = CombineTimePanelProps<DateType> & NormalTimePanelProps<DateType>;

export type PickerPanelProps<DateType> = CommonPickerPanelProps<DateType> & TimePanelProps<DateType>;

// -----------------------------User-----------------------------
export type SharedPickerProps<DateType> = SharedPanelProps<DateType> &
  SharedTriggerProps<DateType> &
  SharedTimeProps<DateType> &
  CommonProps<DateType>;

export type PickerProps<DateType> = {
  generateConfig: GenerateConfig<DateType>;
  picker: PickerType;
  pickerRef?: React.MutableRefObject<PickerRefConfig>;
} & SharedPickerProps<DateType>;

// displayName?: string;

// move to relatived file
