// @universe-design/react/src/shared/components/picker/Picker.tsx
import type { ReactNode } from 'react';
import React, { useState, useRef, useMemo, useCallback } from 'react';
import noop from 'lodash/noop';
import clsx from 'clsx';
import TimeOutlined from '@universe-design/icons-react/TimeOutlined';
import useControlled from '../../../hooks/useControlled';
import { isFunction, getDefaultSizeAbbr, __SCREENSHOT__ } from '../../../util';
import { formatValue, parseValue } from './utils/ui';
import useSyncState from './hooks/useSyncState';
import composeRefs from './../../../util/composeRefs';
import { isEqual } from './utils/date';
import type { PickerContextType } from './Context';
import { PickerContext } from './Context';
import InputWrapper from './Comp/input-wrapper';
import { useComponentSize, useInputGroupSize, useCssScope, useLocale } from '../../../config-provider/hooks';
import useClickOutsideEffect from './hooks/useClickOutsideEffect';
import PickerTrigger from './Comp/PickerTrigger';
import PickerPanel from './Comp/PickerPanel';
import type { PickerProps, PickerRefConfig, MODE, PanelRefProps } from './interface';
import NativeInput from '../Input';

import './style';

// Wrap with class component to enable pass generic with instance method
function InnerPicker<DateType>(props: PickerProps<DateType>) {
  const timeOperationRef = React.useRef<PanelRefProps>({});

  const {
    locale: localeName,
    components: { picker: locale },
  } = useLocale();
  const {
    generateConfig,
    triggerElement,
    style,
    className,
    onKeyDown = noop,
    onMouseDown = noop,
    onMouseUp = noop,
    onClick = noop,
    onClear,
    dropdownClassName,
    popupStyle,
    getPopupContainer,
    suffixIcon = <TimeOutlined />,
    allowClear,
    autoFocus,
    disabled,
    size = 'medium',
    parse,
    picker,
    mode = picker,
    placeholder = locale.timePlaceholder,
    inputReadOnly,
    format = 'HH:mm',
    error,
    onPanelChange,
  } = props;

  function dateToText(date?: DateType) {
    return formatValue(generateConfig, date, {
      format,
      mode: picker,
      locale,
      localeName,
    });
  }
  const defaultSize = useComponentSize();
  const inputGroupSize = useInputGroupSize();
  const cmpSize = getDefaultSizeAbbr(size || inputGroupSize || defaultSize);

  const [open, setOpened] = useControlled<boolean>(props, {
    defaultValuePropName: '',
    valuePropName: 'open',
    onChangePropName: 'onOpenChange',
    defaultValue: __SCREENSHOT__,
  });
  const [focused, setFocused] = useState(false);

  // eslint-disable-next-line react/destructuring-assignment
  const [value, setValue] = useControlled<DateType | undefined>(props);
  const [localMode, setLocalMode] = useSyncState<MODE>(mode);

  /**
   * 当前视图可见的有效date
   */
  const [viewValue, setViewValue] = useSyncState(value);
  const validText = dateToText(viewValue);

  /**
   * 输入框的值
   */
  const [text, setText] = useSyncState<string>(validText);

  /**
   * hover相关的值
   */
  const [hoverDate, setHoverDate] = useState<DateType>();
  const hoverText = dateToText(hoverDate);
  const panelDivRef = useRef<HTMLDivElement>(null);
  const inputDivRef = useRef<HTMLDivElement>(null);
  const inputRef = useRef<HTMLInputElement>(null);
  const triggerElementRef = useRef<HTMLElement>(null);

  useClickOutsideEffect(
    useMemo(() => [panelDivRef, inputDivRef, triggerElementRef], []),
    open,
    useCallback(() => setOpened(false), [setOpened]),
  );

  const cssScope = useCssScope('picker');
  const prefixCls = cssScope();
  const rootCls = clsx(cssScope(), {
    [cssScope(`--size-${cmpSize}`)]: cmpSize,
    [cssScope('--error')]: error,
    [cssScope('--focus')]: focused,
  });

  function confirm(newValue?: DateType) {
    if (!isEqual(generateConfig, value, newValue)) {
      setValue(newValue);
    }

    // 暂时同步value值，不然受控模式下，可能出现问题
    setViewValue(value);
    setText(dateToText(value));

    setHoverDate(undefined);
    setOpened(false);
  }

  function handleKeyDown(e: React.KeyboardEvent<HTMLInputElement>) {
    onKeyDown(e);

    if (e.key !== 'Tab' && !open) {
      setOpened(true);
      return;
    }
    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    e.keyCode === 13 && confirm(viewValue);
  }

  const handleInputChange: React.ChangeEventHandler<HTMLInputElement> = e => {
    setHoverDate(undefined);
    const editText = e.target.value;
    setText(editText);
    const date = parseValue<DateType>(generateConfig, editText, {
      format,
      parse,
      mode,
      locale,
      localeName,
    });

    if (
      date &&
      generateConfig.isValidate(date)
      // && !(typeof disabledDate === 'function' && disabledDate(date))
    ) {
      setViewValue(date);
    }
  };

  const handleClear = () => {
    setValue(undefined);
    // 暂时同步value值，不然受控模式下，可能出现问题
    setViewValue(value);
    setOpened(false);

    onClear && onClear();
  };

  const handleMouseUp: React.MouseEventHandler<HTMLDivElement> = e => {
    if (inputRef.current) {
      inputRef.current.focus();
    }
    setOpened(true);
    onMouseUp(e);
  };

  const handleFocus: React.FocusEventHandler<HTMLInputElement> = e => {
    setFocused(true);
  };

  const handleBlur: React.FocusEventHandler<HTMLInputElement> = e => {
    setFocused(false);
    confirm(viewValue);
  };

  let triggerElementNode: ReactNode;
  if (isFunction(triggerElement)) {
    triggerElementNode = triggerElement({
      text,
      hoverText,
      open,
      value,
    });
  } else {
    triggerElementNode = triggerElement;
  }

  const handlePanelChange = (date: DateType) => {
    setViewValue(date);
    setText(
      formatValue(generateConfig, date, {
        format,
        mode,
        locale,
        localeName,
      }),
    );
    onPanelChange && onPanelChange(date);
  };

  const popupElement = (
    <div onMouseDown={e => e.preventDefault()} ref={panelDivRef}>
      <PickerContext.Provider
        value={
          {
            mode: localMode,
            setMode: setLocalMode,
            value,
            setValue,
            setOpened,
          } as PickerContextType<DateType>
        }
      >
        <PickerPanel
          {...props}
          prefixCls={prefixCls}
          locale={locale}
          value={viewValue}
          operationRef={timeOperationRef}
          onPanelChange={handlePanelChange}
        />
      </PickerContext.Provider>
    </div>
  );

  const renderElement = () => {
    if (triggerElementNode) {
      // 如果用户传递了非 ReactElement 这里会直接报错，导致无法正常渲染
      return React.cloneElement(triggerElementNode as React.ReactElement, {
        ref: composeRefs<HTMLElement>(
          triggerElementRef,
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          (triggerElementNode as any).ref,
        ),
      });
    }
    return (
      <div style={style} ref={inputDivRef} className={rootCls}>
        <InputWrapper
          className={clsx(className, `${prefixCls}-dateInput`)}
          showClear={Boolean(viewValue)}
          onClear={handleClear}
          prefixCls={prefixCls}
          onMouseUp={handleMouseUp}
          onMouseDown={onMouseDown}
          onClick={onClick}
          focused={focused}
          suffixIcon={suffixIcon}
          disabled={disabled}
          allowClear={allowClear}
        >
          <NativeInput
            size={12}
            className={clsx({
              [`${prefixCls}-input`]: true,
              [`${prefixCls}-input-placeholder`]: Boolean(hoverText),
            })}
            placeholder={placeholder}
            disabled={disabled}
            autoFocus={autoFocus}
            ref={inputRef}
            onKeyDown={handleKeyDown}
            onFocus={handleFocus}
            onBlur={handleBlur}
            onChange={handleInputChange}
            readOnly={inputReadOnly || (isFunction(format) && !isFunction(parse))}
            value={hoverText || text}
            data-testid={cssScope('__input')}
          />
        </InputWrapper>
      </div>
    );
  };

  return (
    <PickerTrigger<DateType>
      open={open}
      disabled={disabled}
      setOpened={setOpened}
      dropdownClassName={dropdownClassName}
      popupStyle={popupStyle}
      popupElement={popupElement}
      getPopupContainer={getPopupContainer}
    >
      {renderElement()}
    </PickerTrigger>
  );
}

class Picker<DateType> extends React.Component<PickerProps<DateType>> {
  pickerRef = React.createRef<PickerRefConfig>();

  focus = () => {
    if (this.pickerRef.current) {
      this.pickerRef.current.focus();
    }
  };

  blur = () => {
    if (this.pickerRef.current) {
      this.pickerRef.current.blur();
    }
  };

  render() {
    return (
      <InnerPicker<DateType> {...this.props} pickerRef={this.pickerRef as React.MutableRefObject<PickerRefConfig>} />
    );
  }
}

export default Picker;
