// @universe-design/react/src/portal/stories/homeless.stories.tsx
/** @jsxImportSource @emotion/react */
import * as React from 'react';
import { createPortal } from 'react-dom';
import { Portal } from '..';
import { Dialog, Drawer, Button, Divider } from '../..';
import { useForceUpdate } from '@universe-design/react-utilities';

export default {
  title: 'Components/Portal/临时示例',
  component: Portal,
};

export const Basic = () => (
  <div>
    <span>I am here</span>
    <Portal>This text is portaled at the end of document.body!</Portal>
  </div>
);

export const Nested = () => (
  <Portal>
    This is a portal.
    <Portal>This is a nested portal</Portal>
  </Portal>
);

export const RenderToContainer = () => {
  const ref = React.useRef<any>(null);
  return (
    <>
      <div data-testid="container" ref={ref} />
      <Portal container={ref}>
        <h1 data-testid="heading">Hello world</h1>
      </Portal>
    </>
  );
};

export const RenderToContainerAndNested = () => {
  const ref = React.useRef(null);
  return (
    <div>
      <Portal container={ref}>
        <div>
          Parent: Hey welcome,
          <Portal appendToParentPortal={false}>Child: I am going to document.body</Portal>
        </div>
      </Portal>
      <div style={{ background: 'red' }} ref={ref} />
    </div>
  );
};

export const RealWorldNested = () => {
  const [visible1, setVisible1] = React.useState(false);
  const [visible2, setVisible2] = React.useState(false);
  const [visible3, setVisible3] = React.useState(false);
  return (
    <>
      <Button onClick={() => setVisible1(true)}>打开 Dialog1</Button>
      <Dialog visible={visible1} onCancel={() => setVisible1(false)}>
        <Drawer visible={visible2} onClose={() => setVisible2(false)}>
          <Dialog visible={visible3} onCancel={() => setVisible3(false)} />
          <Button onClick={() => setVisible3(true)}>打开 Dialog2</Button>
        </Drawer>
        <Button onClick={() => setVisible2(true)}>打开 Drawer1</Button>
      </Dialog>
    </>
  );
};

const Frame: React.FC<React.HTMLAttributes<HTMLIFrameElement>> = ({ children, ...props }) => {
  const iframeRef = React.useRef<HTMLIFrameElement | null>(null);
  const forceUpdate = useForceUpdate();
  const mountNode = iframeRef.current?.contentWindow?.document?.body;

  return (
    <iframe
      srcDoc="<!DOCTYPE html><html><head></head><body></body></html>"
      {...props}
      ref={iframeRef}
      onLoad={forceUpdate}
    >
      {mountNode && createPortal(children, mountNode)}
    </iframe>
  );
};

export const ContainerChange = () => {
  const [flag, setFlag] = React.useState(true);

  return (
    <>
      <Button onClick={() => setFlag(!flag)}>点击切换 container</Button>
      <div id="test1" style={{ border: '1px solid purple', width: 100, height: 100 }} />
      <Frame>
        <Portal container={flag ? '#test1' : undefined}>test</Portal>
      </Frame>
    </>
  );
};

export const ParentPortalNone = () => (
  <>
    <Divider text="无 parentPortal 无 iframe" />
    <Portal>无 parentPortal 无 iframe 就渲染到主窗口中</Portal>
  </>
);

const Component1 = <Portal key="1">test</Portal>;
const Component2 = <Frame key="2">{Component1}</Frame>;

const Component = ({ flag }: { flag: boolean }) => (flag ? Component2 : Component1);
export const ParentPortalNoneInIframe = () => {
  const [flag, setFlag] = React.useState<boolean>(true);
  return (
    <>
      <Button onClick={() => setFlag(!flag)}>点击 从逻辑上把 Portal 挪到 iframe 外面</Button>
      <Component flag={flag} />
    </>
  );
};

export const ParentPortal = () => (
  <>
    <Divider text="有 parentPortal" />
    <Portal>
      有 parentPortal 的话<Portal>就挂在 parentPortal 中</Portal>
    </Portal>
  </>
);
