// @universe-design/react/src/pro/timezone-picker/stories/site.stories.tsx
// autogen by ../../../../../scripts/generate-stories.ts

import type { StoryFn as Story } from '@storybook/react';
import C0Basic from '../demo/0.basic';
import C1Basic from '../demo/1.basic';
import C2Localization from '../demo/2.localization';
if (!(C0Basic as Story).storyName) {
  (C0Basic as Story).storyName = '0. 基础用法';
  (C0Basic as Story).parameters = {
    docs: {
      description: {
        story: '时区选择组件。\n\n',
      },
    },
  };
}
if (!(C1Basic as Story).storyName) {
  (C1Basic as Story).storyName = '1. 文本形式';
  (C1Basic as Story).parameters = {
    docs: {
      description: {
        story: '时区文本展示。\n\n',
      },
    },
  };
}
if (!(C2Localization as Story).storyName) {
  (C2Localization as Story).storyName = '2. 本地化';
  (C2Localization as Story).parameters = {
    docs: {
      description: {
        story: '异化时区名称，使产品在不同的地区获得更好的接受度和用户体验。\n\n',
      },
    },
  };
}

export default {
  title: 'Components/TimezonePicker/官网示例',
  tags: ['autodocs'],
};

export { C0Basic, C1Basic, C2Localization };
