// @universe-design/react/src/badge/index.tsx
/* eslint-disable max-lines-per-function */
/* eslint-disable @typescript-eslint/indent */
import MoreOutlined from '@universe-design/icons-react/MoreOutlined';
import clsx from 'clsx';
import React from 'react';
import stylex from '@universe-design/stylex';
import { cssBlockBase, cssFontMedium } from '../stylex';
import { isNonNullable } from '../util';
import composeRefs from '../util/composeRefs';
import { useProxyRef, useControlledState } from '@universe-design/react-utilities';
import { getBadgeRadius, getBorderRadius, isPresetColor, normalizeCssSize, useBadge } from './utils';
import type { BadgeProps, BaseBadgeProps } from './types';

import './style';

const MAX_BADGE_DISPLAY_NUMBER = 999;
const BADGE_FONT_SIZE = 10;

const cssBadgeBadge = stylex.compose(
  cssBlockBase,
  cssFontMedium,
  stylex.create({
    position: 'absolute',
    zIndex: 'auto',
    fontFamily: '"PingFang SC", "Microsoft YaHei", "Myriad Pro", "Hiragino Sans GB", sans-serif', // 关键是 PingFang SC 和 Microsoft YaHei 对应 mac 和 windows
    fontSize: BADGE_FONT_SIZE, // chrome、new edge 使用 12px，其他支持 10px 的浏览器会显示为 10px，包括 safari、firefox、chromium、volcan、lark-native
  }),
);

export const BaseBadge = React.forwardRef<HTMLElement, React.PropsWithChildren<BaseBadgeProps>>((props, ref) => {
  const { children, status, offset, style: styleProp, className, title } = props;
  let { count } = props;

  const { badgeCssScope, shape, position, isHidden, showAsDot, hasStatus } = useBadge(props);

  // Title
  const titleNode = title ?? (typeof count === 'string' || typeof count === 'number' ? String(count) : undefined);

  if (isHidden) {
    return <>{children}</>;
  }

  const hasChild = Boolean(children);
  let style: React.CSSProperties = { ...styleProp };
  let colorClassName = '';
  const badgeBadgeCls = clsx(
    badgeCssScope('__badge'),
    !children && badgeCssScope('-has-no-child'),
    stylex(
      cssBadgeBadge,
      !children &&
        stylex.create({
          transform: 'none',
          position: 'relative',
        }),
    ),
  );

  if (hasChild) {
    /**
     * 📢 定位要用 translate3d，在某些场景下可能出现 badge 内容左右偏移，加这个可以让 badge 独立一层
     */
    if (position === 'topRight') {
      style = {
        top: 0,
        right: 0,
        transform: offset
          ? `translate3d(calc(1 * (${normalizeCssSize(offset[0])})), calc(-1 * (${normalizeCssSize(offset[1])})), 0)`
          : undefined,
        ...style,
      };
    } else if (position === 'topLeft') {
      style = {
        top: 0,
        left: 0,
        transform: offset
          ? `translate3d(calc(-1 * (${normalizeCssSize(offset[0])})), calc(-1 * (${normalizeCssSize(offset[1])})), 0)`
          : undefined,
        ...style,
      };
    } else if (position === 'bottomRight') {
      style = {
        bottom: 0,
        right: 0,
        transform: offset
          ? `translate3d(calc(1 * (${normalizeCssSize(offset[0])})), calc(1 * (${normalizeCssSize(offset[1])})), 0)`
          : undefined,
        ...style,
      };
    } else {
      style = {
        bottom: 0,
        left: 0,
        transform: offset
          ? `translate3d(calc(-1 * (${normalizeCssSize(offset[0])})), calc(1 * (${normalizeCssSize(offset[1])})), 0)`
          : undefined,
        ...style,
      };
    }
  }
  if (showAsDot) {
    // Dot badge, Status badge
    const { colorStyle = 'filled', color, text, size } = props;
    const radius = getBadgeRadius({
      size,
      dot: true,
    });
    style = { width: radius * 2, height: radius * 2, flexShrink: 0, ...style };
    if (color) {
      // color 优先级高于 status
      if (isPresetColor(color)) {
        colorClassName = clsx(badgeCssScope(`-${color}-${colorStyle}`), badgeCssScope(`-${colorStyle}`));
      } else if (colorStyle === 'filled') {
        style = { backgroundColor: color, ...style };
      } else {
        style = {
          backgroundColor: 'transparent',
          border: `1px solid ${color}`,
          ...style,
        };
      }
    } else if (status) {
      colorClassName = clsx(badgeCssScope(`-${status}-${colorStyle}`), badgeCssScope(`-${colorStyle}`));
    } else {
      colorClassName = clsx(badgeCssScope(`-red-${colorStyle}`), badgeCssScope(`-${colorStyle}`));
    }

    if (!hasStatus) {
      return (
        <span className={clsx(badgeCssScope(), badgeCssScope('-dot'), className)}>
          {children}
          <sup
            ref={ref}
            title={titleNode}
            className={clsx(badgeBadgeCls, badgeCssScope(`-${position}`), colorClassName)}
            style={style}
          />
        </span>
      );
    } else {
      return (
        <span className={clsx(badgeCssScope(), badgeCssScope('-dot-inline'), className)}>
          <span ref={ref} className={clsx(badgeBadgeCls, colorClassName)} style={style} />
          {isNonNullable(text) && <span className={clsx(badgeCssScope('__text'))}>{text}</span>}
        </span>
      );
    }
  } else {
    // Text badge
    let type = 'text' as 'text' | 'icon';
    const {
      overflowCount = MAX_BADGE_DISPLAY_NUMBER,
      color = 'red',
      hack10px = false,
      textBadgeSidecar: TextBadgeSidecar,
    } = props;

    // 处理 Sidecar 逻辑
    const supClassName = clsx(badgeBadgeCls, badgeCssScope(`-${position}`));

    const isCountOverflow = typeof count === 'number' && count > overflowCount;

    if (TextBadgeSidecar && typeof count !== 'object' && !isCountOverflow) {
      return (
        <span className={clsx(badgeCssScope(), className)}>
          {children}
          <TextBadgeSidecar color={color} className={supClassName} style={style} content={String(count)} />
        </span>
      );
    }

    if (isPresetColor(color)) {
      colorClassName = badgeCssScope(`-${color}-filled`);
    } else {
      style = { backgroundColor: color, ...style };
    }
    if (shape === 'circle' || String(count).length === 1) {
      type = 'icon';
    }
    if (isCountOverflow) {
      count = <MoreOutlined className={badgeCssScope('__badge__more')} />;
    }

    return (
      <span className={clsx(badgeCssScope(), badgeCssScope(`-${type}`), className)}>
        {children}
        <sup ref={ref} title={titleNode} className={clsx(supClassName, colorClassName)} style={style}>
          <div
            className={clsx(badgeCssScope('__badge__content'), hack10px && badgeCssScope('__badge__content-hack-10px'))}
          >
            {count}
          </div>
        </sup>
      </span>
    );
  }
});

export const Badge = React.forwardRef<HTMLElement, React.PropsWithChildren<BadgeProps>>((props, ref) => {
  const {
    children,
    offset: offsetProp,
    position,
    contentType,
    childCornerRadius: childCornerRadiusProp,
    ...restProps
  } = props;

  const { showAsDot } = useBadge(props);

  const shouldGetChildrenBorderRadius = !isNonNullable(childCornerRadiusProp) && !isNonNullable(offsetProp);
  const [childCornerRadius, setChildCornerRadius] = useControlledState(0, childCornerRadiusProp);
  const childRef = useProxyRef<HTMLElement | null>(null, node => {
    if (node && node instanceof HTMLElement && shouldGetChildrenBorderRadius) {
      setChildCornerRadius(getBorderRadius(node, position));
    }
  });
  const mergedChildRef = React.useMemo(() => composeRefs(childRef, (children as any)?.ref), [childRef, children]);
  const badgeRadius = getBadgeRadius(props);
  const badgeCenterPoint = { x: badgeRadius, y: badgeRadius };
  const tangencyPoint = {
    x: (1 - Math.SQRT1_2) * childCornerRadius,
    y: (1 - Math.SQRT1_2) * childCornerRadius,
  };
  // Offset
  const offset = React.useMemo(() => {
    let _offset: [number | string, number | string] = [
      -tangencyPoint.x + badgeCenterPoint.x,
      -tangencyPoint.y + badgeCenterPoint.y,
    ];
    if (contentType === 'text' && showAsDot) {
      _offset = ['2px + 100%', 0];
    } else if (contentType === 'text') {
      _offset = ['2px + 100%', '50%'];
    } else if (isNonNullable(offsetProp)) {
      _offset = offsetProp;
    }
    return _offset;
  }, [tangencyPoint.x, tangencyPoint.y, badgeCenterPoint.x, badgeCenterPoint.y, contentType, showAsDot, offsetProp]);

  // 如果是有 offsetProp 或者 children 没法 cloneELement 或者不需要获取子元素圆角的时候直接返回 base badge
  if (!React.isValidElement(children) || !shouldGetChildrenBorderRadius) {
    return (
      <BaseBadge {...restProps} ref={ref} offset={offset} position={position}>
        {children}
      </BaseBadge>
    );
  }
  return (
    <BaseBadge {...restProps} ref={ref} offset={offset} position={position}>
      {React.cloneElement(children as React.ReactElement<React.RefAttributes<HTMLElement | null>>, {
        ref: mergedChildRef,
      })}
    </BaseBadge>
  );
});

export * from './SVGTextBadge';
export * from './CanvasTextBadge';
export * from './types';
