// @universe-design/react/src/badge/types.ts
import type React from 'react';
import type {
  DefaultExtraSizeType,
  DefaultSizeType,
  LiteralUnion,
  PositionType,
  PresetColorType,
  PresetStatusColorType,
} from '../util';

export type TextBadgeSidecarComponentType = React.ComponentType<{
  color: string;
  className: string;
  style: React.CSSProperties;
  content: string;
}>;

type BadgePropColor = LiteralUnion<PresetColorType, string>;

/**
 * @public
 */
export interface BaseBadgeProps {
  /**
   * 徽标元素的 `style` 属性透传，可覆盖原来的样式
   */
  style?: React.CSSProperties;
  /**
   * 徽标元素的 `className` 属性透传
   */
  className?: string;
  /**
   * 对于 `filled` 类型为背景颜色，对于 `outlined` 则是边框颜色
   * @defaultValue `red`
   */
  color?: BadgePropColor;
  /**
   * 展示的数字，大于 `overflowCount` 时显示为 `...`
   */
  count?: React.ReactNode;
  /**
   * 不展示数字，只有一个小红点
   * @defaultValue false
   */
  dot?: boolean;
  /**
   * 设置状态点的位置偏移，以徽标组件的中心向外为正方向
   */
  offset?: [number | string, number | string];
  /**
   * 徽标位置
   * @defaultValue `topRight`
   */
  position?: Extract<PositionType, 'topLeft' | 'bottomLeft' | 'topRight' | 'bottomRight'>;
  /**
   * 展示封顶的数字值，只当 `count` 属性是 `number` 时有效
   * @defaultValue 999
   */
  overflowCount?: number;
  /**
   * 当数值为 0 时，是否展示徽标
   * @defaultValue false
   */
  showZero?: boolean;
  /**
   * 设置小圆点的大小。如果传入数字则作为徽标小圆点的半径值。仅在点状徽标下有效
   * @defaultValue `medium`
   */
  size?: Extract<DefaultSizeType | DefaultExtraSizeType, 'small' | 'medium' | 'large'> | number;
  /**
   * 设置徽标为状态点，布局为横向内联
   */
  status?: PresetStatusColorType | 'waiting' | 'missing';
  /**
   * 在设置了 `status` 的前提下有效，设置状态点的文本
   */
  text?: React.ReactNode;
  /**
   * 设置鼠标放在状态点上时显示的文字
   */
  title?: HTMLElement['title'];
  /**
   * 点状徽标的上色方式，分为 `filled`、`outlined` 和 `borderer` 三种。
   * @defaultValue `filled`
   */
  colorStyle?: 'filled' | 'outlined' | 'borderer';
  /**
   * 文字徽标可用，标注徽标形状，圆形或圆角矩形
   * @defaultValue `round`
   */
  shape?: 'round' | 'circle';
  /**
   * 是否用 12px + 0.8333 zoom hack 10 号字，在 chrome 和 edge 浏览器可以选择性开启
   * @defaultValue false
   */
  hack10px?: boolean;
  /**
   * @internal
   * 可替换的纯文字徽标组件，默认使用 HTML 实现，可选择 `SVGTextBadgeSidecar` 作为替换选项
   */
  textBadgeSidecar?: TextBadgeSidecarComponentType;
}

/**
 * @public
 */
export interface BadgeProps extends BaseBadgeProps {
  /**
   * 徽标包裹内容的类型，会影响徽标的定位
   * @defaultValue `default`
   */
  contentType?: 'text' | 'default';
  /**
   * 内容元素的圆角值。不传入时会根据子元素样式获取圆角值并计算徽标位置
   */
  childCornerRadius?: number;
}
