// @universe-design/react/src/table/test-demo/stretch.tsx
import * as React from 'react';
import { EmptyNegativeError } from '@universe-design/illustration-react';
import type { TableProps } from '@universe-design/react';
import { Empty, Table } from '@universe-design/react';

interface TableData {
  key: string;
  name: string;
  column3: string;
  children?: TableData[];
}

const columns = [
  {
    title: 'Name',
    dataIndex: 'name',
    width: '200px',
  },
  {
    title: 'Column 2',
    dataIndex: 'column2',
  },
  {
    title: 'Action',
    key: 'action',
    render: () => <a>Delete</a>,
  },
];

const generateData = (limit: number, level: number, prefix: number) => {
  const data: TableData[] = [];
  const levelStack = [];
  while (level) {
    levelStack.push(level);
    level--;
  }
  const levelPrefix = levelStack.join('-');
  for (let i = 1; i <= limit; i++) {
    data.push({
      key: `${prefix}-${levelPrefix}-${i}`,
      name: `John Brown ${levelPrefix}`,
      column3: `New York No. ${i} Lake Park`,
    });
  }
  return data;
};

export type TableDemoStretchProps = Pick<TableProps<TableData>, 'style' | 'sticky' | 'pagination'> & {
  rowCount: number;
};

export default ({ rowCount, ...props }: TableDemoStretchProps) => {
  const tableData = generateData(rowCount, 1, 0);

  return (
    <Table
      stretch
      columns={columns}
      dataSource={tableData}
      renderEmpty={<Empty image={<EmptyNegativeError />} title="居中显示" />}
      {...props}
    />
  );
};
