// @universe-design/react/lib/button/ButtonBase.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ButtonBase = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _isNil = _interopRequireDefault(require("lodash/isNil"));
var _svgs = require("../_assets/svgs");
var _hooks = require("../config-provider/hooks");
var _util = require("../util");
var _composeRefs = _interopRequireDefault(require("../util/composeRefs"));
var _useLoading = require("../loading/useLoading");
require("./style");
var _markAsDisabledableComp = require("../shared/markAsDisabledableComp");
// type ReactButtonProps = React.DetailedHTMLProps<
//   React.ButtonHTMLAttributes<HTMLButtonElement>,
//   HTMLButtonElement
// >;

/**
 * Deprecated 组件的分类：https://bytedance.feishu.cn/docs/doccngkxeLeLStbebeToeKkPW7b#
 * Breaking change
 * - 重写 type，type 表示按钮类型，每种类型有自己的限制条件，如:
 *  - filled 限制背景改变，color 不变
 *  - outlined 限制背不变
 *  - ghost 限制背景透明
 * - 重写样式，降低样式的层叠优先级
 * - 修正 ButtonLink, LinkButton 的实现
 * - 去掉部分 Button 负 margin
 * - 去掉 info danger 属性，统一使用 color
 *  - info => color="primary"
 *  - danger => color="danger"
 * - 去掉 ghost 属性，ghost => type="ghost"
 *
 */

/**
 * @public
 */

/**
 * 支持原生`button`的所有事件和属性。
 * @public
 * @docgen
 */

const customizeSizeSet = new Set(['icon']);
function getSizeClass(type) {
  return customizeSizeSet.has(type) ? `${type}-size` : 'size';
}
const ButtonCmp = /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
  const {
    prefixCls: customizePrefixCls,
    className,
    children,
    shape,
    htmlType = 'button',
    type = 'outlined',
    size: customizeSize,
    color = 'default',
    selected = false,
    loading: loadingProp = false,
    delay,
    icon,
    endIcon,
    block = false,
    noMinWidth = false,
    onClick,
    danger,
    info,
    ...rest
  } = props;
  process.env.NODE_ENV !== "production" ? (0, _util.warning)((0, _isNil.default)(props.autoBlur), (0, _util.deprecatedWarning)('button', 'autoBlur')) : void 0;
  const cssScope = (0, _hooks.useCssScope)('button');
  const prefixCls = customizePrefixCls || cssScope();
  const [loading] = (0, _useLoading.useLoading)({
    loading: loadingProp,
    delay
  });
  const size = (0, _hooks.useComponentSize)();
  const buttonRef = (0, _react.useRef)();
  const mergedRef = (0, _react.useMemo)(() => (0, _composeRefs.default)(ref, buttonRef), [ref, buttonRef]);
  const sizeValue = (0, _util.getSizeAbbr)(customizeSize || size || 'medium');
  let variant = type;
  if (type === 'primary') {
    variant = 'filled';
  }
  if (type === 'default') {
    variant = 'outlined';
  }
  let usedColor = color;
  if (usedColor === 'default' && danger) {
    usedColor = 'danger';
  }
  if (usedColor === 'default' && info) {
    usedColor = 'primary';
  }
  const classes = (0, _clsx.default)(prefixCls, {
    [cssScope(`--${variant}`)]: variant,
    [cssScope(`--${variant}-${usedColor}`)]: usedColor,
    [cssScope(`--${variant}-${usedColor}-loading`)]: loading,
    [cssScope(`--${variant}-${usedColor}-selected`)]: selected,
    [cssScope('--loading')]: loading,
    [cssScope('--selected')]: selected,
    [cssScope(`--${shape}`)]: shape,
    [cssScope(`--${getSizeClass(variant)}-${sizeValue}`)]: sizeValue,
    [cssScope('--disabled')]: rest.disabled,
    [cssScope('--block')]: block,
    [cssScope('--autowidth')]: noMinWidth
  }, className);
  let leftIcon = icon;
  let rightIcon = endIcon;
  if (leftIcon && children) {
    leftIcon = /*#__PURE__*/_react.default.createElement("span", {
      className: (0, _clsx.default)(cssScope('__icon-inline'), cssScope('__icon-inline-start'))
    }, leftIcon);
  }
  if (rightIcon && children) {
    rightIcon = /*#__PURE__*/_react.default.createElement("span", {
      className: (0, _clsx.default)(cssScope('__icon-inline'), cssScope('__icon-inline-end'))
    }, rightIcon);
  }
  if (loading && !rest.disabled) {
    leftIcon = /*#__PURE__*/_react.default.createElement("span", {
      className: (0, _clsx.default)(cssScope('__loading-icon'), cssScope('__icon-inline'), children && cssScope('__icon-inline-start'))
    }, /*#__PURE__*/_react.default.createElement(_svgs.SpinLoadingSvg, null));
  }
  const handleClick = (0, _react.useCallback)(e => {
    if (loading || rest.disabled) {
      e.stopPropagation();
      e.preventDefault();
    } else {
      onClick && onClick(e);
    }
  }, [loading, rest.disabled, onClick]);
  if (rest.href !== undefined) {
    return /*#__PURE__*/_react.default.createElement("a", (0, _extends2.default)({}, rest, {
      ref: mergedRef,
      onClick: handleClick,
      className: classes
      // disabled={disabled}
    }), leftIcon, children, rightIcon);
  }
  return /*#__PURE__*/ /* eslint-disable react/button-has-type */_react.default.createElement("button", (0, _extends2.default)({}, rest, {
    type: htmlType,
    ref: mergedRef,
    onClick: handleClick,
    className: classes
  }), leftIcon, children, rightIcon)
  /* eslint-enable react/button-has-type */;
});
const ButtonBase = exports.ButtonBase = /* #__PURE__*/(0, _markAsDisabledableComp.markAsDisabledableComp)(ButtonCmp);
ButtonBase.__UD_BUTTON = true;
if (process.env.NODE_ENV !== "production") {
  ButtonBase.displayName = 'ButtonBase';
}
