// @universe-design/react/lib/button/Link.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Link = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _svgs = require("../_assets/svgs");
var _hooks = require("../config-provider/hooks");
var _markAsDisabledableComp = require("../shared/markAsDisabledableComp");
/**
 * Deprecated
 * 这是一个根据 Universse Design 设计，做的一个 Button 抽象，它强大而灵活
 */

const LinkCmp = /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
  const {
    prefixCls: customizePrefixCls,
    className,
    children,
    loading = false,
    disabled = false,
    icon: iconProp,
    onClick,
    ...rest
  } = props;
  const cssScope = (0, _hooks.useCssScope)('link');
  const prefixCls = customizePrefixCls || cssScope();
  const aRef = (0, _react.useRef)();
  (0, _react.useImperativeHandle)(ref, () => aRef.current);
  const classes = (0, _clsx.default)(prefixCls, {
    [cssScope('--loading')]: loading && !disabled,
    [cssScope('--disabled')]: disabled
  }, className);
  let icon = iconProp;
  if (loading && !disabled) {
    icon = /*#__PURE__*/_react.default.createElement(_svgs.SpinLoadingSvg, null);
  }
  const handleClick = (0, _react.useCallback)(e => {
    if (loading || disabled) {
      e.stopPropagation();
      e.preventDefault();
    } else {
      onClick && onClick(e);
    }
  }, [loading, disabled, onClick]);
  return /*#__PURE__*/_react.default.createElement("a", (0, _extends2.default)({}, rest, {
    ref: a => {
      if (a) {
        aRef.current = a;
      }
    },
    className: classes,
    onClick: handleClick
  }), icon, children);
});
const Link = exports.Link = /* #__PURE__*/(0, _markAsDisabledableComp.markAsDisabledableComp)(LinkCmp);
Link.__UD_LINK = true;
Link.__UD_BUTTON = true;
if (process.env.NODE_ENV !== "production") {
  Link.displayName = 'Link';
}
