// @universe-design/react/lib/tree/hooks/checkStrategy.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.strategyMap = exports.checkStrategyShowParent = exports.checkStrategyShowChild = exports.checkStrategyShowAll = void 0;
var _interface = require("../interface");
var _util = require("../util");
var _getCheckedCache = require("./getCheckedCache");
// 尽可能用指针来减少 key 太长带来的内存问题

/**
 * show all
 */
const checkStrategyShowAll = exports.checkStrategyShowAll = {
  cache2CheckedSet(doublyTreeNodeMap, cache) {
    const checkedSet = {};
    for (const [k, v] of cache.entries()) {
      checkedSet[k] = v === _interface.CheckedType.Checked;
    }
    return checkedSet;
  },
  checkedSet2Cache(doublyTreeNodeMap, checkedSet) {
    const checkedKeys = Object.entries(checkedSet).filter(_ref => {
      let [, v] = _ref;
      return v === true;
    }).map(_ref2 => {
      let [x] = _ref2;
      return x;
    });
    return (0, _getCheckedCache.getCheckedCache)({
      doublyTreeNodeMap,
      checkedKeys,
      checkStrictly: false
    });
  }
};

/**
 * show parent
 * 如果所有的 children 都选中则只显示 parent
 */
const checkStrategyShowParent = exports.checkStrategyShowParent = {
  ...checkStrategyShowAll,
  cache2CheckedSet(doublyTreeNodeMap, cache) {
    const checkedSet = {};
    // 补充不存在的 value
    for (const [key, checkedType] of cache) {
      if (checkedType === _interface.CheckedType.Checked) {
        var _node$parent, _node$parent$children;
        checkedSet[key] = true;
        const node = doublyTreeNodeMap[key];
        const allSiblingsChecked = node === null || node === void 0 ? void 0 : (_node$parent = node.parent) === null || _node$parent === void 0 ? void 0 : (_node$parent$children = _node$parent.children) === null || _node$parent$children === void 0 ? void 0 : _node$parent$children.every(_node => cache.get(_node.__node__.value) === _interface.CheckedType.Checked
        //  || isCheckedDisabled(_node.__node__),
        );
        checkedSet[key] = !allSiblingsChecked;
      } else {
        checkedSet[key] = false;
      }
    }
    return checkedSet;
  }
};

/**
 * show children
 * 如果所有的 children 都只显示 parent
 */
const checkStrategyShowChild = exports.checkStrategyShowChild = {
  ...checkStrategyShowAll,
  cache2CheckedSet(doublyTreeNodeMap, cache) {
    const checkedSet = {};
    for (const [key, checkedType] of cache.entries()) {
      const node = doublyTreeNodeMap[key];
      // 可能 node 不存在，此时仍然将选择态保留
      const isChild = !node || !(0, _util.isNonLeaf2)(node);
      checkedSet[key] = checkedType === _interface.CheckedType.Checked && isChild;
    }
    return checkedSet;
  }
};
const strategyMap = exports.strategyMap = {
  [_interface.ShowCheckedStrategy.SHOW_ALL]: checkStrategyShowAll,
  [_interface.ShowCheckedStrategy.SHOW_CHILD]: checkStrategyShowChild,
  [_interface.ShowCheckedStrategy.SHOW_PARENT]: checkStrategyShowParent
};
