// @universe-design/react/lib/tree/hooks/useTree.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useTree = exports.TreeProvider = void 0;
var React = _interopRequireWildcard(require("react"));
var _reactUtilities = require("@universe-design/react-utilities");
var _common = require("../common");
var _interface = require("../interface");
var _TreeContext = require("./TreeContext");
var _useTreeSelect = require("./useTreeSelect");
var _useTreeCheck = require("./useTreeCheck");
var _useTreeExpand = require("./useTreeExpand");
var _checkStrategy = require("./checkStrategy");
var _util = require("../util");
var _configProvider = require("../../config-provider");
/**
 * @docgen
 */

const defaultFilterTreeNode = () => true;
const useTree = (props, ref) => {
  const locale = (0, _configProvider.useLocale)();
  const childrenToTreeData = React.useMemo(() => (0, _common.treeChildren2TreeNodes)(props.children), [props.children]);
  const {
    treeData = childrenToTreeData,
    defaultExpandedKeys = [],
    expandedKeys,
    onExpand,
    defaultSelectedKeys = [],
    selectedKeys,
    onSelect,
    filterTreeNode = defaultFilterTreeNode,
    showChecker = false,
    checkable = false,
    selectable = true,
    defaultExpandAll = false,
    multiple,
    showCheckedStrategy = _interface.ShowCheckedStrategy.SHOW_ALL,
    switcherIcon,
    labelRenderer,
    loadData,
    onLoad,
    loadedKeys: loadedKeysProp,
    searchValue,
    notFoundContent = locale.global.notFoundContent,
    // virtual scrolling
    virtual = true,
    height,
    nodeRenderer
  } = props;
  const cssScope = (0, _common.useTreeCssScope)();
  const [loadedKeys, setLoadedKeys] = (0, _reactUtilities.useControlledState)([], loadedKeysProp);

  // ======== 创建一个带索引的 tree node 双向链表 ========
  const {
    doublyTreeNodeList,
    doublyTreeNodeMap
  } = React.useMemo(() => (0, _util.genDoublyTreeNodeMap)(treeData), [treeData]);
  const {
    finalTreeNodeList,
    filteredTreeNodes,
    expandedMap,
    toggleExpanded,
    finalTreeNodeListRef
  } = (0, _useTreeExpand.useTreeExpand)({
    doublyTreeNodeMap,
    doublyTreeNodeList,
    defaultExpandAll,
    defaultExpandedKeys,
    expandedKeys,
    onExpand,
    filterTreeNode,
    searchValue
  });
  const treeCheckCtx = (0, _useTreeCheck.useTreeCheck)({
    doublyTreeNodeMap,
    checkStrategy: _checkStrategy.strategyMap[showCheckedStrategy],
    ...(props.checkStrictly === true ? {
      ...props,
      checkStrictly: true
    } : {
      ...props,
      checkStrictly: false
    })
  });
  const treeSelectCtx = (0, _useTreeSelect.useTreeSelect)({
    doublyTreeNodeMap,
    defaultSelectedKeys,
    selectedKeys,
    onSelect,
    multiple
  });

  // =========================== List ===========================
  const listRef = React.useRef(null);
  const scrollIntoView = React.useCallback((reference, scrollConfig) => {
    var _doublyTreeNodeMap$ke, _doublyTreeNodeMap$ke2;
    if (!listRef.current) {
      return;
    }
    if (typeof reference === 'number') {
      listRef.current.scrollTo({
        index: reference,
        ...scrollConfig
      });
      return;
    }
    const {
      key
    } = reference;
    const parentKey = (_doublyTreeNodeMap$ke = doublyTreeNodeMap[key]) === null || _doublyTreeNodeMap$ke === void 0 ? void 0 : (_doublyTreeNodeMap$ke2 = _doublyTreeNodeMap$ke.__parent__) === null || _doublyTreeNodeMap$ke2 === void 0 ? void 0 : _doublyTreeNodeMap$ke2.value;
    if (parentKey && !expandedMap[parentKey]) {
      toggleExpanded(parentKey);
    }
    /**
     * 这里需要一个异步处理：
     * 等待节点展开、下一次 render 之后，才能拿到更新后的 finalTreeNodeList
     * */
    setTimeout(() => {
      const index = finalTreeNodeListRef.current.findIndex(n => n.key === key);
      if (index !== -1) {
        var _listRef$current;
        (_listRef$current = listRef.current) === null || _listRef$current === void 0 ? void 0 : _listRef$current.scrollTo({
          index,
          ...scrollConfig
        });
      }
    });
  }, [expandedMap, listRef, finalTreeNodeListRef, toggleExpanded, doublyTreeNodeMap]);
  React.useImperativeHandle(ref, () => ({
    toggleExpanded,
    scrollIntoView
  }), [scrollIntoView, toggleExpanded]);
  return React.useMemo(() => ({
    loadData,
    onLoad,
    loadedKeys,
    selectable,
    checkable,
    setLoadedKeys,
    rootCssScope: cssScope,
    nodeRenderer,
    showChecker,
    expandedMap,
    toggleExpanded,
    switcherIcon,
    finalTreeNodeList,
    labelRenderer,
    filteredTreeNodes,
    ...treeCheckCtx,
    ...treeSelectCtx,
    height,
    virtual,
    listRef,
    searchValue,
    notFoundContent
  }), [checkable, cssScope, expandedMap, filteredTreeNodes, finalTreeNodeList, labelRenderer, loadData, loadedKeys, nodeRenderer, onLoad, selectable, setLoadedKeys, showChecker, switcherIcon, toggleExpanded, treeCheckCtx, treeSelectCtx, height, virtual, searchValue, notFoundContent]);
};
exports.useTree = useTree;
const TreeProvider = exports.TreeProvider = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    children,
    ...props
  } = _ref;
  const treeCtxValue = useTree(props, ref);
  return /*#__PURE__*/React.createElement(_TreeContext.TreeContextProvider, {
    value: treeCtxValue
  }, children);
});
