// @universe-design/react/lib/tree/hooks/useTreeCheck.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useTreeCheck = useTreeCheck;
var _reactUtilities = require("@universe-design/react-utilities");
var _react = _interopRequireDefault(require("react"));
var _util = require("../../util");
var _interface = require("../interface");
var _util2 = require("../util");
var _TreeCheckedContext = require("./TreeCheckedContext");
var _getCheckedCache = require("./getCheckedCache");
function useTreeCheck(props) {
  let {
    skipped: skippedOpt = false
  } = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  const {
    doublyTreeNodeMap,
    checkStrategy,
    checkStrictly,
    defaultCheckedKeys,
    checkedKeys: checkedKeysProp
  } = props;

  // 这段逻辑保证单独使用 useTreeCheck 也一定没有问题
  const treeValueContext = (0, _TreeCheckedContext.useTreeCheckedContext)();
  const skipped = Boolean(treeValueContext) || skippedOpt;

  // eslint-disable-next-line prefer-const
  let [checkedKeys, setCheckedKeys] = (0, _reactUtilities.useControlledState)(defaultCheckedKeys !== null && defaultCheckedKeys !== void 0 ? defaultCheckedKeys : checkStrictly ? {
    halfChecked: [],
    checked: []
  } : [], checkedKeysProp, {
    isControlledOverride: 'checkedKeys' in props
  });
  let checkedCache = _react.default.useMemo(() => {
    if (skipped) {
      return new Map();
    }
    if (checkStrictly) {
      const cache = new Map();
      const {
        checked,
        halfChecked
      } = checkedKeys;
      checked.forEach(k => cache.set(k, _interface.CheckedType.Checked));
      halfChecked.forEach(k => cache.set(k, _interface.CheckedType.Indeterminate));
      Object.keys(doublyTreeNodeMap).forEach(key => {
        if (cache.get(key) === undefined) {
          cache.set(key, _interface.CheckedType.Unchecked);
        }
      });
      return cache;
    } else {
      const keys = (0, _util.genArray)(checkedKeys);
      return checkStrategy.checkedSet2Cache(doublyTreeNodeMap, (0, _util2.array2Set)(keys));
    }
  }, [checkStrategy, checkStrictly, checkedKeys, doublyTreeNodeMap, skipped]);
  const checkedSet = _react.default.useMemo(() => {
    if (skipped) {
      return {};
    }
    if (checkStrictly) {
      const r = {};
      for (const [k, v] of checkedCache.entries()) {
        if (v === _interface.CheckedType.Checked) {
          r[k] = true;
        }
      }
      return r;
    }
    return checkStrategy.cache2CheckedSet(doublyTreeNodeMap, checkedCache);
  }, [skipped, checkStrictly, checkStrategy, doublyTreeNodeMap, checkedCache]);
  if (treeValueContext) {
    ({
      checkedCache,
      setCheckedKeys
    } = treeValueContext);
  }

  /**
   * 切换 checked 状态
   */
  const toggleChecked = _react.default.useCallback(key => {
    const node = doublyTreeNodeMap[key];
    if (!node || (0, _util2.isCheckedDisabled)(node.__node__)) {
      return;
    }
    const checkedType = checkedCache.get(node.__node__.value);
    let keyToRemove;
    let keyToAdd;
    if (checkedType === _interface.CheckedType.Checked) {
      keyToRemove = node.key;
    } else {
      keyToAdd = node.key;
    }
    if (props.checkStrictly) {
      const checkedCache = (0, _getCheckedCache.getCheckedCache)({
        doublyTreeNodeMap,
        checkedKeys: checkedKeys,
        keyToAdd,
        keyToRemove,
        checkStrictly: true
      });
      const newCheckedKeys = {
        checked: [],
        halfChecked: []
      };
      for (const [k, v] of checkedCache.entries()) {
        if (v === _interface.CheckedType.Checked) {
          newCheckedKeys.checked.push(k);
        } else if (v === _interface.CheckedType.Indeterminate) {
          newCheckedKeys.halfChecked.push(k);
        }
      }
      setCheckedKeys(newCheckedKeys, () => (0, _util.isFunction)(props.onCheck) && props.onCheck(newCheckedKeys, {
        checked: checkedType !== _interface.CheckedType.Checked,
        key
      }));
    } else {
      const checkedCache = (0, _getCheckedCache.getCheckedCache)({
        doublyTreeNodeMap,
        checkedKeys: checkedKeys,
        keyToAdd,
        keyToRemove,
        checkStrictly: false
      });
      const checkedMap = checkStrategy.cache2CheckedSet(doublyTreeNodeMap, checkedCache);
      const newKeys = Object.keys(checkedMap).filter(k => checkedMap[k]);
      setCheckedKeys(newKeys, () => (0, _util.isFunction)(props.onCheck) && props.onCheck(newKeys, {
        checked: checkedType !== _interface.CheckedType.Checked,
        key
      }));
    }
  }, [doublyTreeNodeMap, checkedCache, props, checkedKeys, setCheckedKeys, checkStrategy]);
  return _react.default.useMemo(() => ({
    checkedCache,
    checkedSet,
    setCheckedKeys,
    toggleChecked
  }), [checkedCache, checkedSet, setCheckedKeys, toggleChecked]);
}
