// @universe-design/react/lib/upload/uploadFile/CardFile.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CardFile = void 0;
var _ug24l6d = _interopRequireDefault(require("./../../.stylex/ug24l6d"));
var _uidkx = _interopRequireDefault(require("./../../.stylex/uidkx17"));
var _u1u1p7mj = _interopRequireDefault(require("./../../.stylex/u1u1p7mj"));
var _u8rtfz = _interopRequireDefault(require("./../../.stylex/u8rtfz1"));
var _ui763hu = _interopRequireDefault(require("./../../.stylex/ui763hu"));
var _ubo3qzx = _interopRequireDefault(require("./../../.stylex/ubo3qzx"));
var _u5jxv = _interopRequireDefault(require("./../../.stylex/u5jxv71"));
var _clsx = _interopRequireDefault(require("clsx"));
var _react = _interopRequireDefault(require("react"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _UploadProgress = require("./UploadProgress");
var _utils = require("../utils");
var _FileContext = require("./FileContext");
var _useFile = require("./useFile");
var _FileName = require("./FileName");
require("../style/card-file.less");
var _util = require("../../util");
//.u1jhi6b8{flex-grow:0;}
//.ui64zpg{flex-shrink:0;}
//.u1euv43f{position:absolute;}
//.u5gq2j6{bottom:0px;}
//.u9j3dip{left:8px;}
//.u112f5iw{right:8px;}
//.u10pi13n{position:relative;}
const CardFile = fileProps => {
  const {
    style,
    nameRenderer
  } = fileProps;
  const fileCtxValue = (0, _useFile.useFile)(fileProps);
  const {
    classNames,
    file,
    errorNode,
    retryNode,
    removeNode,
    downloadNode,
    previewNode,
    iconRender,
    rootPrefixCls,
    progressProps
  } = fileCtxValue;
  process.env.NODE_ENV !== "production" ? (0, _util.warning)(!['downloading', 'downloadFailed', 'downloaded'].includes(file.status), (0, _util.devWarning)('Upload', `file.status=${file.status} does not support for CardFile.`)) : void 0;
  return /*#__PURE__*/_react.default.createElement(_FileContext.FileContextProvider, {
    value: fileCtxValue
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(classNames, `${rootPrefixCls}--card`),
    style: style
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)(`${rootPrefixCls}__preview`, (0, _stylex.default)({
      "props": {
        "qhf8xq": _u5jxv.default
      }
    }))
  }, function () {
    const previewFile = file;
    const previewUrl = previewFile.thumbUrl || previewFile.url;
    if ((0, _utils.isImage)(previewFile) && previewUrl && previewFile.status === 'success') {
      return /*#__PURE__*/_react.default.createElement("div", {
        className: `${rootPrefixCls}-preview-img`
      }, /*#__PURE__*/_react.default.createElement("img", {
        src: previewUrl
      }));
    }
    return /*#__PURE__*/_react.default.createElement("div", {
      className: `${rootPrefixCls}-preview-icon`
    }, iconRender(previewFile));
  }(), /*#__PURE__*/_react.default.createElement(_UploadProgress.UploadProgress, {
    file: file,
    progressProps: progressProps,
    prefixCls: rootPrefixCls,
    className: (0, _stylex.default)({
      "props": {
        "qhf8xq": _u1u1p7mj.default,
        "B5kzvoi": _u8rtfz.default,
        "oyh7mz": _ui763hu.default,
        "j35jbq": _ubo3qzx.default
      }
    })
  })), /*#__PURE__*/_react.default.createElement("div", {
    className: `${rootPrefixCls}--card__info`
  }, /*#__PURE__*/_react.default.createElement("span", {
    className: `${rootPrefixCls}--card__info-icon`
  }, iconRender(file)), /*#__PURE__*/_react.default.createElement("div", {
    className: `${rootPrefixCls}--card__info-name`
  }, /*#__PURE__*/_react.default.createElement(_FileName.FileName, {
    file: file,
    nameRenderer: nameRenderer
  })), /*#__PURE__*/_react.default.createElement("span", {
    className: (0, _clsx.default)(`${rootPrefixCls}--card__info-operations`, (0, _stylex.default)({
      "props": {
        "Bh6795r": _ug24l6d.default,
        "Bnnss6s": _uidkx.default
      }
    }))
  }, errorNode, previewNode, downloadNode, retryNode, removeNode))));
};
exports.CardFile = CardFile;
