// @universe-design/react/lib/upload/uploadFile/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CardFile", {
  enumerable: true,
  get: function () {
    return _CardFile.CardFile;
  }
});
Object.defineProperty(exports, "PictureFile", {
  enumerable: true,
  get: function () {
    return _PictureFile.PictureFile;
  }
});
Object.defineProperty(exports, "TextFile", {
  enumerable: true,
  get: function () {
    return _TextFile.TextFile;
  }
});
Object.defineProperty(exports, "TextFileAction", {
  enumerable: true,
  get: function () {
    return _TextFile.TextFileAction;
  }
});
exports.UploadTextFile = exports.UploadSimpleTextFile = exports.UploadPictureFile = exports.UploadCardFile = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _CardFile = require("./CardFile");
var _PictureFile = require("./PictureFile");
var _TextFile = require("./TextFile");
var _useUploadFile = require("./useUploadFile");
const UploadTextFile = props => {
  const uploadFileProps = (0, _useUploadFile.useUploadFile)(props);
  return /*#__PURE__*/React.createElement(_TextFile.TextFile, uploadFileProps);
};
exports.UploadTextFile = UploadTextFile;
const SimpleTextFile = props => /*#__PURE__*/React.createElement(_TextFile.TextFile, (0, _extends2.default)({}, props, {
  isSimple: true
}));
const UploadSimpleTextFile = props => {
  const uploadFileProps = (0, _useUploadFile.useUploadFile)(props);
  return /*#__PURE__*/React.createElement(SimpleTextFile, uploadFileProps);
};
exports.UploadSimpleTextFile = UploadSimpleTextFile;
const UploadCardFile = props => {
  const uploadFileProps = (0, _useUploadFile.useUploadFile)(props);
  return /*#__PURE__*/React.createElement(_CardFile.CardFile, uploadFileProps);
};
exports.UploadCardFile = UploadCardFile;
const UploadPictureFile = props => {
  const uploadFileProps = (0, _useUploadFile.useUploadFile)(props);
  return /*#__PURE__*/React.createElement(_PictureFile.PictureFile, uploadFileProps);
};
exports.UploadPictureFile = UploadPictureFile;
