// @universe-design/react/lib/upload/uploadFile/TextFile.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TextFile = TextFile;
exports.TextFileAction = void 0;
var _u12o6wmt = _interopRequireDefault(require("./../../.stylex/u12o6wmt"));
var _clsx = _interopRequireDefault(require("clsx"));
var _react = _interopRequireDefault(require("react"));
var _configProvider = require("../../config-provider");
var _textOverflow = require("../../shared/components/text-overflow");
var _UploadProgress = require("./UploadProgress");
var _utils = require("../utils");
var _useFile = require("./useFile");
var _FileName = require("./FileName");
var _FileContext = require("./FileContext");
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _react2 = require("@universe-design/stylex/react");
require("../style/text-file.less");
//.u1k6fduh{cursor:pointer;}
/** @jsxRuntime classic */
/** @jsx jsx */
/** @jsxFrag React.Fragment */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const DownloadAction = () => {
  const {
    downloadNode
  } = (0, _FileContext.useFileContext)();
  return downloadNode;
};
const RetryAction = () => {
  const {
    retryNode
  } = (0, _FileContext.useFileContext)();
  return retryNode;
};
const PreviewAction = () => {
  const {
    previewNode
  } = (0, _FileContext.useFileContext)();
  return previewNode;
};
const RemoveAction = () => {
  const {
    removeNode
  } = (0, _FileContext.useFileContext)();
  return removeNode;
};
const TextFileAction = exports.TextFileAction = {
  Download: DownloadAction,
  Retry: RetryAction,
  Preview: PreviewAction,
  Remove: RemoveAction
};

/**
 * 与 Upload 无关的 File 组件
 */

const FileStatus = _ref => {
  let {
    file,
    prefixCls
  } = _ref;
  const locale = (0, _configProvider.useLocaleReceiver)('upload');
  if (file.status === 'waiting') {
    return (0, _react2.jsx)(_textOverflow.TextOverflow, {
      as: "p",
      className: `${prefixCls}-info-subtitle`,
      lineClamp: 1
    }, locale.waiting);
  }
  if ((0, _utils.isError)(file)) {
    const defaultFailedReason = file.status === 'error' ? locale.failed : locale.downloadFailed;
    const failedReason = file.failedReason || defaultFailedReason;
    return (0, _react2.jsx)(_textOverflow.TextOverflow, {
      as: "p",
      className: `${prefixCls}-info-error`,
      lineClamp: 2
    }, failedReason);
  }
  if (file.status === 'uploading' || file.status === 'downloading') {
    return (0, _react2.jsx)(_textOverflow.TextOverflow, {
      as: "p",
      className: `${prefixCls}-info-subtitle`,
      lineClamp: 1
    }, file.status === 'downloading' && `${locale.downloading} `, (0, _utils.formatFileSize)(file.loaded, locale), " / ", (0, _utils.formatFileSize)(file.size, locale));
  }
  return (0, _react2.jsx)(_textOverflow.TextOverflow, {
    as: "p",
    className: `${prefixCls}-info-subtitle`,
    lineClamp: 1
  }, (0, _utils.formatFileSize)(file.size, locale));
};
const defaultAction = (0, _react2.jsx)(_react.default.Fragment, null, (0, _react2.jsx)(TextFileAction.Preview, {
  key: "preview"
}), (0, _react2.jsx)(TextFileAction.Retry, {
  key: "retry"
}), (0, _react2.jsx)(TextFileAction.Download, {
  key: "download"
}), (0, _react2.jsx)(TextFileAction.Remove, {
  key: "remove"
}));
function TextFile(props) {
  const {
    style,
    isSimple,
    nameRenderer,
    action = defaultAction
  } = props;
  const fileCtxValue = (0, _useFile.useFile)(props);
  const {
    file,
    errorNode,
    iconRender,
    rootPrefixCls,
    onPreview,
    progressProps,
    classNames
  } = fileCtxValue;
  const operationAreaRef = _react.default.useRef(null);
  const previewable = ['success', 'downloading', 'downloaded', 'downloadFailed'].includes(file.status);
  return (0, _react2.jsx)(_FileContext.FileContextProvider, {
    value: fileCtxValue
  }, (0, _react2.jsx)("div", {
    className: (0, _clsx.default)(classNames, isSimple ? `${rootPrefixCls}--simple-text` : `${rootPrefixCls}--text`),
    onClick: function (e) {
      var _operationAreaRef$cur;
      if (!previewable) {
        return;
      }
      const {
        target
      } = e;
      // 防止操作按钮触发 item preview
      if ((target instanceof HTMLElement || target instanceof SVGElement) && (e.target === operationAreaRef.current || (_operationAreaRef$cur = operationAreaRef.current) !== null && _operationAreaRef$cur !== void 0 && _operationAreaRef$cur.contains(target))) {
        return;
      }
      onPreview === null || onPreview === void 0 ? void 0 : onPreview();
    },
    style: style,
    sx: [previewable && {
      "props": {
        "Bceei9c": _u12o6wmt.default
      }
    }]
  }, (0, _react2.jsx)("div", {
    className: `${rootPrefixCls}-container`
  }, (0, _react2.jsx)("div", {
    className: `${rootPrefixCls}-icon`
  }, iconRender === null || iconRender === void 0 ? void 0 : iconRender(file)), (0, _react2.jsx)("div", {
    className: `${rootPrefixCls}-info`
  }, (0, _react2.jsx)(_FileName.FileName, {
    file: file,
    nameRenderer: nameRenderer,
    className: `${rootPrefixCls}-info-name`
  }), !isSimple && (0, _react2.jsx)(FileStatus, {
    file: file,
    prefixCls: rootPrefixCls
  })), (0, _react2.jsx)("div", {
    className: `${rootPrefixCls}-operations`,
    ref: operationAreaRef
  }, isSimple && errorNode, action)), (0, _react2.jsx)(_UploadProgress.UploadProgress, {
    file: file,
    progressProps: progressProps,
    prefixCls: rootPrefixCls
  })));
}
