// @universe-design/react/lib/upload/uploadFile/UploadProgress.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UploadProgress = UploadProgress;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _clsx = _interopRequireDefault(require("clsx"));
var _react = _interopRequireDefault(require("react"));
var _progress = require("../../progress");
function UploadProgress(_ref) {
  let {
    prefixCls,
    file,
    progressProps,
    className,
    ...htmlProps
  } = _ref;
  if (file.status === 'uploading' || file.status === 'waiting' || file.status === 'downloading') {
    return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
      className: (0, _clsx.default)(`${prefixCls}-progress`, className)
    }, htmlProps), /*#__PURE__*/_react.default.createElement(_progress.Progress, (0, _extends2.default)({
      type: "line",
      showInfo: false,
      percent: file.percent,
      strokeWidth: 4
    }, progressProps)));
  }
  return null;
}
