// @universe-design/react/lib/upload/uploadFile/useFile.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OperationIconComp = OperationIconComp;
exports.useFile = useFile;
var _ulvoz0o = _interopRequireDefault(require("./../../.stylex/ulvoz0o"));
var React = _interopRequireWildcard(require("react"));
var _DeleteTrashOutlined = _interopRequireDefault(require("@universe-design/icons-react/DeleteTrashOutlined"));
var _DownloadOutlined = _interopRequireDefault(require("@universe-design/icons-react/DownloadOutlined"));
var _RefreshOutlined = _interopRequireDefault(require("@universe-design/icons-react/RefreshOutlined"));
var _VisibleOutlined = _interopRequireDefault(require("@universe-design/icons-react/VisibleOutlined"));
var _WarningFilled = _interopRequireDefault(require("@universe-design/icons-react/WarningFilled"));
var _stylex = _interopRequireDefault(require("@universe-design/stylex"));
var _clsx = _interopRequireDefault(require("clsx"));
var _hooks = require("../../config-provider/hooks");
var _utils = require("../utils");
var _tooltip = require("../../tooltip");
var _button = require("../../button");
var _getFileIcon = _interopRequireDefault(require("../utils/getFileIcon"));
var _FileContext = require("./FileContext");
//.u114py8l{color:@function-danger-content-default;}
function OperationIconComp(_ref) {
  let {
    icon,
    handleClick,
    tooltip,
    className,
    style,
    color
  } = _ref;
  const {
    file,
    rootPrefixCls: prefixCls
  } = (0, _FileContext.useFileContext)();
  if (typeof icon === 'function') {
    icon = icon(file);
  }
  if (!icon) {
    return null;
  }
  return /*#__PURE__*/React.createElement(_tooltip.Tooltip, {
    title: tooltip
  }, /*#__PURE__*/React.createElement(_button.IconButton, {
    className: (0, _clsx.default)(`${prefixCls}-operation-icon`, className),
    icon: icon,
    onClick: () => handleClick === null || handleClick === void 0 ? void 0 : handleClick(file),
    style: style,
    size: "small",
    color: color
  }));
}
function useFile(props) {
  const {
    file: fileProp,
    prefixCls,
    className
  } = props;
  const file = React.useMemo(() => ({
    status: 'success',
    ...fileProp
  }), [fileProp]);
  const defaultFileProps = React.useMemo(() => ({
    disabled: false,
    progress: {
      strokeWidth: 4,
      showInfo: false
    },
    showRemoveIcon: true,
    showDownloadIcon: true,
    showPreviewIcon: true,
    showErrorIcon: true,
    showRetryIcon: true,
    removeIcon: /*#__PURE__*/React.createElement(_DeleteTrashOutlined.default, null),
    onPreview: () => (0, _utils.internalPreview)(file),
    previewIcon: /*#__PURE__*/React.createElement(_VisibleOutlined.default, null),
    onDownload: () => (0, _utils.internalDownload)(file),
    downloadIcon: /*#__PURE__*/React.createElement(_DownloadOutlined.default, null),
    retryIcon: /*#__PURE__*/React.createElement(_RefreshOutlined.default, null),
    errorIcon: /*#__PURE__*/React.createElement(_WarningFilled.default, {
      className: (0, _stylex.default)({
        "props": {
          "sj55zd": _ulvoz0o.default
        }
      })
    })
  }), [file]);
  const mergedFileProps = {
    // 默认值优先级最低
    ...defaultFileProps,
    // props 优先级最高
    ...props
  };
  const {
    showDownloadIcon,
    downloadIcon,
    showErrorIcon,
    showPreviewIcon,
    showRemoveIcon,
    showRetryIcon,
    errorIcon,
    previewIcon,
    progress,
    disabled,
    removeIcon,
    retryIcon,
    onRemove,
    onRetry,
    onDownload,
    onPreview,
    iconRender: iconRenderProp
  } = mergedFileProps;
  const rootPrefixCls = (0, _hooks.useCssScope)('upload-item', prefixCls)();
  const locale = (0, _hooks.useLocaleReceiver)('upload');
  const status = file.status || 'success';
  const downloadNode = React.useMemo(() => {
    if (!showDownloadIcon || status !== 'success') {
      return null;
    }
    return /*#__PURE__*/React.createElement(OperationIconComp, {
      icon: downloadIcon,
      handleClick: onDownload,
      tooltip: locale.download
    });
  }, [downloadIcon, locale.download, onDownload, showDownloadIcon, status]);
  const errorNode = React.useMemo(() => {
    if (!showErrorIcon || status !== 'error') {
      return null;
    }
    const failedReason = file.failedReason || locale.failed;
    return /*#__PURE__*/React.createElement(OperationIconComp, {
      icon: errorIcon,
      tooltip: failedReason
    });
  }, [errorIcon, file, locale.failed, showErrorIcon, status]);
  const previewable = showPreviewIcon && ['success', 'downloaded'].includes(status);
  const previewNode = React.useMemo(() => {
    if (!previewable || !previewIcon) {
      return null;
    }
    return /*#__PURE__*/React.createElement(OperationIconComp, {
      icon: previewIcon,
      tooltip: locale.preview,
      handleClick: onPreview
    });
  }, [locale.preview, onPreview, previewable, previewIcon]);
  const retryNode = React.useMemo(() => {
    if (!showRetryIcon || status !== 'error' || disabled || !retryIcon) {
      return null;
    }
    return /*#__PURE__*/React.createElement(OperationIconComp, {
      icon: retryIcon,
      tooltip: locale.uploadAgain,
      handleClick: onRetry,
      className: `${rootPrefixCls}-operation-retry`,
      color: "primary"
    });
  }, [disabled, locale.uploadAgain, onRetry, retryIcon, showRetryIcon, status, rootPrefixCls]);
  const removeNode = React.useMemo(() => {
    if (!showRemoveIcon || disabled || file.status === 'downloading') {
      return null;
    }
    return /*#__PURE__*/React.createElement(OperationIconComp, {
      icon: removeIcon,
      tooltip: locale.delete,
      handleClick: onRemove
    });
  }, [disabled, file, locale.delete, onRemove, removeIcon, showRemoveIcon]);
  const classNames = (0, _clsx.default)(className, rootPrefixCls, `${rootPrefixCls}--${file.status}`, (0, _utils.isError)(file) ? `${rootPrefixCls}--error` : `${rootPrefixCls}--normal`);
  return {
    downloadNode,
    errorNode,
    removeNode,
    previewNode,
    retryNode,
    showPreviewIcon,
    showErrorIcon,
    iconRender: function (f) {
      var _iconRenderProp;
      return (_iconRenderProp = iconRenderProp === null || iconRenderProp === void 0 ? void 0 : iconRenderProp(f)) !== null && _iconRenderProp !== void 0 ? _iconRenderProp : (0, _getFileIcon.default)(f);
    },
    onRemove,
    onPreview,
    file,
    disabled,
    rootPrefixCls,
    progressProps: progress,
    classNames
  };
}
