// @universe-design/react/lib/upload/uploadFile/useUploadFile.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useUploadFile = void 0;
var _react = _interopRequireDefault(require("react"));
var _UploadContext = require("../hooks/UploadContext");
var _utils = require("../utils");
const useUploadFile = props => {
  const {
    file,
    onRemove: onRemoveProp,
    onRetry: onRetryProp,
    onPreview: onPreviewProp,
    onDownload: onDownloadProp,
    disabled: disabledProp,
    ...resetProps
  } = props;
  const onPreview = _react.default.useCallback(() => onPreviewProp ? onPreviewProp() : (0, _utils.internalPreview)(file), [file, onPreviewProp]);
  const onDownload = _react.default.useCallback(() => onDownloadProp ? onDownloadProp() : (0, _utils.internalDownload)(file), [file, onDownloadProp]);
  const {
    onRetry: onRetryCtx,
    disabled,
    onRemove: onRemoveCtx
  } = (0, _UploadContext.useUploadContext)();
  const onRemove = _react.default.useCallback(() => {
    Promise.resolve(typeof onRemoveProp === 'function' ? onRemoveProp() : true).then(ret => {
      // Prevent removing file
      if (ret === false) {
        return;
      }
      onRemoveCtx(file);
    });
  }, [file, onRemoveCtx, onRemoveProp]);
  const onRetry = _react.default.useCallback(() => {
    onRetryCtx(file);
    onRetryProp === null || onRetryProp === void 0 ? void 0 : onRetryProp();
  }, [file, onRetryCtx, onRetryProp]);
  return {
    onPreview,
    onRemove,
    onRetry,
    onDownload,
    disabled: disabledProp !== null && disabledProp !== void 0 ? disabledProp : disabled,
    file,
    ...resetProps
  };
};
exports.useUploadFile = useUploadFile;
