// @universe-design/react/src/select/context/SelectContext.ts
import type { RefObject } from 'react';
import React from 'react';
import type { SelectAction, InternalSelectOptionConfig, InternalSelectProps, SelectValue } from '../internal.interface';
import type { RefOptionListProps, SelectProps } from '../interface';
import type { UseSelectDataMapReturnType } from './useSelectDataMap';
import type { Required } from 'utility-types';
import type { useSelectOverlayVisible, useSelectInputValue, useSelectFocus, useSelectInput } from './internal';

/**
 * @public
 * @docgen
 * @typeParam V - value
 * @typeParam OP - option
 */
export interface SelectContextType<V extends SelectValue, OP extends InternalSelectOptionConfig<V>>
  extends Pick<SelectProps<V>, 'tabIndex' | 'size'>,
    Omit<UseSelectDataMapReturnType<V, OP>, 'setValueSet' | 'optionMap'>,
    Omit<ReturnType<typeof useSelectFocus>, 'focusRef'>,
    ReturnType<typeof useSelectInput>,
    ReturnType<typeof useSelectOverlayVisible>,
    ReturnType<typeof useSelectInputValue>,
    Required<
      Pick<
        InternalSelectProps<V, OP>,
        | 'disabled'
        | 'showSearch'
        | 'onSearch'
        | 'tagRenderer'
        | 'id'
        | 'notFoundContent'
        | 'noOptionsContent'
        | 'dropdownRender'
        | 'loading'
      >,
      'disabled' | 'showSearch' | 'tagRenderer' | 'notFoundContent' | 'noOptionsContent' | 'dropdownRender' | 'loading'
    > {
  listRef: RefObject<RefOptionListProps<V>>;
  select: (_value: V) => void;
  deselect: (_value: V) => void;
  toggle: (
    _value: V,
    option?: {
      closeOverlay?: boolean;
    },
  ) => void;
  set: (
    _value: V[],
    option?: {
      closeOverlay?: boolean;
      action?: SelectAction;
    },
  ) => void;
  clear: () => void;
  deselectLast: () => void;
  cssScope: (cls?: string) => string;
  selectorCssScope: (cls?: string) => string;
  dropdownCssScope: (cls?: string) => string;
  multiple: boolean;
  selectorLabelRenderer: (value: V) => React.ReactNode;
  inputting: boolean;
  setInputting: React.Dispatch<boolean>;
  /**
   * 是否对选项内容做文字超长省略
   */
  textOverflow: boolean;
}

export const SelectContext = React.createContext<
  SelectContextType<SelectValue, InternalSelectOptionConfig<SelectValue>>
>({} as any);
if (__DEV__) {
  SelectContext.displayName = 'SelectContext';
}
