// @universe-design/react/src/select/context/useSelectFocus.ts
import type { FocusEvent, FocusEventHandler } from 'react';
import { useCallback, useRef, useState } from 'react';
import { useConstant } from '@universe-design/react-utilities';
import { useEffectOnce, useEventCallback } from '../../hooks';
import { isFunction } from '../../util';
import { SelectSearchAction } from '../internal.interface';
import type { useSelectOverlayVisible } from './internal';
import type { useSelectInputValue } from './useSelectInputValue';

function callDelay<A extends any[]>(
  cb: (...args: A) => void,
  duration = 250,
): { call: (...args: A) => void; cancel: () => void } {
  let timeoutID: number | null = null;

  const cancel = () => {
    if (timeoutID !== null) {
      window.clearTimeout(timeoutID);
      timeoutID = null;
    }
  };

  const call = (...args: A) => {
    cancel();
    timeoutID = window.setTimeout(() => cb(...args), duration);
  };

  return { call, cancel };
}

/**
 * use select focus management
 * 使用 input 的 focus 状态来代替整个组件的 focus 状态
 */
export function useSelectFocus({
  disabled,
  autoFocus,
  onFocus,
  onBlur,
  setOverlayVisible,
  overlayVisible,
  setInputValue,
  setInputting,
}: {
  disabled: boolean;
  autoFocus: boolean;
  onFocus?: (event: FocusEvent<HTMLElement>) => void;
  onBlur?: (event: FocusEvent<HTMLElement>) => void;
  setInputting: React.Dispatch<boolean>;
} & ReturnType<typeof useSelectOverlayVisible> &
  Pick<ReturnType<typeof useSelectInputValue>, 'setInputValue'>) {
  const _setOverlayVisible = useEventCallback(setOverlayVisible);
  const _onBlur = useEventCallback(onBlur);
  const inputRef = useRef<HTMLInputElement>(null);
  const focusRef = useRef<boolean>(false);
  const [focusState, setFocusState] = useState(false);
  const { call: callCloseOverlayDelay, cancel: cancelDelayCloseOverlay } = useConstant(() =>
    callDelay((e: FocusEvent<HTMLElement>) => {
      if (isFunction(_onBlur)) {
        _onBlur(e);
      }
      _setOverlayVisible(false);
      setInputValue('', { action: SelectSearchAction.Blur });
    }),
  );

  /**
   * container focus 事件
   *
   * 接受一种不带 event onInputFocus，在 dropdown 点击的时候触发，避免 dropdown 关闭
   */
  const onContainerFocus = useCallback(
    (e: FocusEvent<HTMLElement>) => {
      cancelDelayCloseOverlay();
      if (isFunction(onFocus) && !focusRef.current && e) {
        onFocus(e);
      }
      setFocusState(true);
      focusRef.current = true;
    },
    [onFocus, cancelDelayCloseOverlay],
  );

  const onContainerBlur: FocusEventHandler<HTMLElement> = useCallback(
    e => {
      // 这段放在最前面是为了直接从 dropdown input blur 时 select 可以得到正确的 focus 样式
      setFocusState(false);
      focusRef.current = false;
      if (e.target !== inputRef.current) {
        return;
      }
      // 我们有一个逻辑是点击任何不是 input 的都要重新聚焦 input，这个时候 input 会有短暂的 blur 状态，所以我们需要区分正常 blur 和这种短时间的 blur
      callCloseOverlayDelay(e);
    },
    [callCloseOverlayDelay],
  );

  useEffectOnce(() => {
    // init focus，如果初始 overlay visible 或者 auto focus，则组件为 focus 状态
    if (autoFocus || overlayVisible) {
      setFocusState(true);
      setInputting(true);
      inputRef.current?.focus();
    }
  });

  const imperativeFocus = useCallback<HTMLOrSVGElement['focus']>(
    options => {
      if (disabled) {
        return;
      }
      // 用户手动触发 focus 时，设置为输入中
      setInputting(true);
      inputRef.current?.focus(options);
    },
    [disabled, setInputting],
  );

  const blur = useCallback(() => {
    if (disabled) {
      return;
    }
    inputRef.current?.blur();
  }, [disabled]);

  return {
    focus: imperativeFocus,
    blur,
    focusState,
    inputRef,
    focusRef,
    onContainerFocus,
    onContainerBlur,
    cancelDelayCloseOverlay,
  };
}
