// @universe-design/react/src/select/context/useSelectInput.ts
import * as React from 'react';
import { isFunction, isNonNullable } from '../../util';
import { KeyboardKey } from '@universe-design/react-utilities';
import type { RefOptionListProps, SelectProps } from '../interface';
import type { InternalSelectOptionConfig, SelectValue } from '../internal.interface';
import { SelectSearchAction } from '../internal.interface';
import type { UseSelectDataMapReturnType, useSelectInputValue, useSelectOverlayVisible } from './internal';

function getSeparatedContent(text: string, tokens?: string[]): string[] | null {
  if (!tokens || !tokens.length) {
    return null;
  }

  let match = false;

  function separate(str: string, [token, ...restTokens]: string[]): string[] {
    if (!token) {
      return [str];
    }

    const list = str.split(token);
    match = match || list.length > 1;

    return list
      .reduce((prevList, unitStr) => [...prevList, ...separate(unitStr, restTokens)], [] as string[])
      .filter(unit => unit);
  }

  const list = separate(text, tokens);
  return match ? list : null;
}

export function useSelectInput<V extends SelectValue, OP extends InternalSelectOptionConfig<V>>({
  multiple,
  tokenSeparators,
  inputValue,
  setInputValue,
  set,
  deselectLast,
  setOverlayVisible,
  listRef,
  onKeyDown,
  onKeyUp,
  overlayVisible,
  setInputting,
  getOption,
}: {
  multiple: boolean;
  set: (
    _value: V[],
    option?: {
      closeOverlay?: boolean;
    },
  ) => void;
  deselectLast: () => void;
  listRef: React.RefObject<RefOptionListProps<V>>;
  setInputting: React.Dispatch<boolean>;
} & Pick<UseSelectDataMapReturnType<V, OP>, 'getOption'> &
  ReturnType<typeof useSelectInputValue> &
  Pick<SelectProps<V>, 'tokenSeparators' | 'onKeyDown' | 'onKeyUp'> &
  Pick<ReturnType<typeof useSelectOverlayVisible>, 'overlayVisible' | 'setOverlayVisible'>) {
  // When paste come, ignore next onChange
  const pastedTextRef = React.useRef<string | null>(null);
  const compositionStatusRef = React.useRef<boolean>(false);

  const tokenWithEnter = React.useMemo(
    () => (tokenSeparators || []).some(tokenSeparator => ['\n', '\r\n'].includes(tokenSeparator)),
    [tokenSeparators],
  );

  const onInputChange: React.ChangeEventHandler<HTMLInputElement> = e => {
    let {
      target: { value: searchText },
    } = e;

    // Pasted text should replace back to origin content
    if (tokenWithEnter && pastedTextRef.current && /[\r\n]/.test(pastedTextRef.current)) {
      // CRLF will be treated as a single space for input element
      const replacedText = pastedTextRef.current.replace(/\r\n/g, ' ').replace(/[\r\n]/g, ' ');
      searchText = searchText.replace(replacedText, pastedTextRef.current);
    }

    pastedTextRef.current = null;

    const patchLabels: string[] | null = compositionStatusRef.current
      ? null
      : getSeparatedContent(searchText, tokenSeparators);

    if (!multiple) {
      /* empty */
    } else if (patchLabels) {
      searchText = '';
      set(
        patchLabels
          .map(label => {
            const item = getOption(label as V);
            return item?.value;
          })
          .filter(isNonNullable),
      );
      return;
    }

    setInputValue(searchText, { action: SelectSearchAction.Input });
  };

  const onInputCompositionStart = () => {
    compositionStatusRef.current = true;
  };

  const onInputCompositionEnd = () => {
    compositionStatusRef.current = false;
  };

  const onInputPaste: React.ClipboardEventHandler = e => {
    const { clipboardData } = e;
    const value = clipboardData.getData('text');

    pastedTextRef.current = value;
  };

  // ============================ Keyboard ============================
  const onInternalKeyDown: React.KeyboardEventHandler<Element> = e => {
    isFunction(onKeyDown) && onKeyDown(e);

    if (overlayVisible && listRef.current) {
      listRef.current.onKeyDown(e);
    }

    // 允许 overlay component 阻止默认行为
    if (e.defaultPrevented) {
      return;
    }

    setInputting(true);

    if (multiple && e.key === KeyboardKey.Backspace) {
      // DeleteTag
      if (inputValue === '') {
        deselectLast();
      }
    }

    if (
      [
        // 禁用浏览器 input 的光标默认行为
        KeyboardKey.Up,
        KeyboardKey.Down,
        KeyboardKey.ArrowUp,
        KeyboardKey.ArrowDown,
        // 禁用浏览器 enter 触发 click 事件，当 trigger element 为 button 时会有这种情况出现
        KeyboardKey.Enter,
      ].includes(e.key as KeyboardKey)
    ) {
      e.preventDefault();
    }

    // Enter 事件处理不在这里，因为 Enter 的关闭逻辑比较难统一
    // ESC 关闭
    if (e.key === KeyboardKey.Escape) {
      // overlay 操作关闭 overlay
      setOverlayVisible(false);
      return;
    }

    // 禁用一些不需要在 input 这边操作的键，其他键都可以触发 overlay 打开
    if (
      ![
        KeyboardKey.Shift, // 组合键
        KeyboardKey.Tab, // 浏览器默认行为切换 input 选中态
        KeyboardKey.Escape,
      ].includes(e.key as KeyboardKey)
    ) {
      setOverlayVisible(true);
    }
  };

  // KeyUp
  const onInternalKeyUp: React.KeyboardEventHandler<Element> = e => {
    isFunction(onKeyUp) && onKeyUp(e);

    if (overlayVisible && listRef.current) {
      listRef.current.onKeyUp(e);
    }
  };

  return {
    inputValue,
    onInputChange,
    onInputCompositionStart,
    onInputCompositionEnd,
    onInputPaste,
    onInternalKeyDown,
    onInternalKeyUp,
    compositionStatusRef,
  };
}
