// @universe-design/react/src/select/context/useSelectDataMap.ts
import * as React from 'react';
import { useControlledState, useConstant } from '@universe-design/react-utilities';
import type { UseControlledStateSetState } from '@universe-design/react-utilities';
import { genArray } from '../../util';
import { getOption } from '../common';
import type { InternalSelectOptionConfig, InternalSelectProps, SelectValue } from '../internal.interface';

export function useSelectDataMap<V extends SelectValue, OP extends InternalSelectOptionConfig<V>>({
  defaultValue,
  value: valueProp,
  options,
  valueInProps,
}: Pick<InternalSelectProps<V, OP>, 'defaultValue' | 'value' | 'options'> & {
  valueInProps: boolean;
}): UseSelectDataMapReturnType<V, OP> {
  const selectedCache = useConstant<Map<V, OP>>(() => new Map());

  /**
   * 记录 optionMap 用于方便索引
   */
  const { optionMap } = React.useMemo(
    () =>
      options.reduce(
        ({ optionMap: preOptMap }, cur) => {
          preOptMap.set(cur.value, cur);
          return { optionMap: preOptMap };
        },
        {
          optionMap: new Map() as Map<V, OP>,
        },
      ),
    [options],
  );

  const updateCache = React.useCallback(
    (valueSet: Set<V>) => {
      for (const o of valueSet) {
        selectedCache.set(o, getOption(o, optionMap, selectedCache));
      }
    },
    [optionMap, selectedCache],
  );

  const [valueSet, setValueSet] = useControlledState(new Set(genArray(defaultValue)), new Set(genArray(valueProp)), {
    onChange: updateCache,
    isControlledOverride: valueInProps,
  });

  // 用 useMemo 降低一下 render 触发该流程的频率
  React.useMemo(() => {
    // 每次 render 把 value 值对应的 options 存进 selectedCache
    updateCache(valueSet);
  }, [updateCache, valueSet]);

  return React.useMemo(
    () => ({
      valueSet,
      setValueSet,
      optionMap,
      getOption: v => getOption(v, optionMap, selectedCache),
    }),
    [optionMap, selectedCache, setValueSet, valueSet],
  );
}

export interface UseSelectDataMapReturnType<V extends SelectValue, OP extends InternalSelectOptionConfig<V>> {
  valueSet: Set<V>;
  setValueSet: UseControlledStateSetState<Set<V>>;
  optionMap: Map<V, OP>;
  getOption: (value: V) => OP;
}
