// @universe-design/react/src/select/context/useSelection.tsx
import identity from 'lodash/identity';
import React, { useCallback } from 'react';
import { useCssScope, useLocale } from '../../config-provider/hooks';
import { useDebounce, useEventCallback } from '../../hooks';
import { isFunction } from '../../util';
import { defaultTagRenderer } from '../SelectorTag';
import type { RefOptionListProps } from '../interface';
import type {
  InternalSelectOptionConfig,
  InternalSelectProps,
  SelectAction,
  SelectGeneratorConfig,
  SelectValue,
} from '../internal.interface';
import { SelectActionType, SelectSearchAction } from '../internal.interface';
import type { SelectContextType } from './SelectContext';
import { useSelectFocus, useSelectInput, useSelectInputValue, useSelectOverlayVisible } from './internal';
import { useSelectDataMap } from './useSelectDataMap';

function selectReducer<V extends SelectValue, OP extends InternalSelectOptionConfig<V>>(
  option: OP | undefined,
  multiple: boolean,
  prevState: Set<V>,
): { nextState: Set<V> } {
  if (!option) {
    return { nextState: prevState };
  }
  if (prevState.has(option.value)) {
    return { nextState: prevState };
  }
  if (multiple) {
    const nextState = new Set(prevState);
    nextState.add(option.value);
    return {
      nextState,
    };
  }
  return {
    nextState: new Set<V>([option.value]),
  };
}

function deselectReducer<V extends SelectValue>(key: V, prevState: Set<V>): { nextState: Set<V> } {
  const nextState = new Set(prevState);
  nextState.delete(key);
  return {
    nextState,
  };
}

/**
 * 判断一个 option 是否为 disable 状态
 * @param option
 * @param globalDisabled
 */
export function isDisabledOption<V extends SelectValue>(
  option: InternalSelectOptionConfig<V>,
  globalDisabled: boolean,
) {
  return globalDisabled || Boolean(option.disabled);
}

/**
 * getPermittedOptions
 * 从输入 key list 里筛选出可以操作的 option keys 列表
 */
function getPermittedOptionKeys<V extends SelectValue>(
  keys: SelectValue[],
  globalDisabled: boolean,
  optionMap: Map<SelectValue, InternalSelectOptionConfig<V>>,
) {
  return keys.filter(key => {
    const opt = optionMap.get(key);
    if (!opt) {
      return !globalDisabled;
    }
    if (isDisabledOption(opt, globalDisabled)) {
      return false;
    }
    return true;
  });
}

// eslint-disable-next-line max-lines-per-function
export function useSelection<V extends SelectValue, OP extends InternalSelectOptionConfig<V>>(
  props: InternalSelectProps<V, OP>,
  config: SelectGeneratorConfig<V, OP> = {
    dropdownCssScope: 'select',
    selectorCssScope: 'select',
  },
): SelectContextType<V, OP> {
  const locale = useLocale();
  const {
    open,
    onDropdownVisibleChange,
    value: valueProp,
    defaultValue,
    onClear,
    onSelect,
    onDeselect,
    onSearch,
    onFocus,
    onBlur,
    searchValue,
    tokenSeparators,
    tabIndex,
    disabled = false,
    autoFocus = false,
    defaultOpen = false,
    autoClearSearchValue = true,
    options,
    onKeyDown,
    onKeyUp,
    size,
    tagRender: tagRenderProp,
    tagRenderer: tagRendererProp,
    id,
    notFoundContent = locale.global.notFoundContent,
    dropdownRender = identity,
    loading = false,
    dropdownMatchSelectWidth,
  } = props;

  const multiple = props.mode === 'multiple';

  // multiple default showSearch
  const { showSearch = multiple, noOptionsContent = showSearch ? null : locale.global.noOptions } = props;

  const useSelectDataMapResult = useSelectDataMap<V, OP>({
    value: valueProp,
    defaultValue,
    options,
    valueInProps: 'value' in props,
  });
  const { valueSet, setValueSet, optionMap, getOption } = useSelectDataMapResult;

  const { inputValue, setInputValue } = useSelectInputValue({
    searchValue,
    onSearch,
  });

  const listRef = React.useRef<RefOptionListProps<V>>(null);

  //  ========================== User Inputting ==========================

  const [inputting, setInputting] = React.useState(false);

  // ========================== Overlay Visible ==========================

  const { overlayVisible, setOverlayVisible } = useSelectOverlayVisible({
    disabled,
    defaultOpen,
    open,
    onDropdownVisibleChange,
    setInputting,
  });

  // ========================== Focus / Blur ==========================

  const { focusState, inputRef, onContainerFocus, onContainerBlur, focus, blur, cancelDelayCloseOverlay } =
    useSelectFocus({
      autoFocus,
      onBlur,
      onFocus,
      disabled,
      overlayVisible,
      setOverlayVisible,
      setInputValue,
      setInputting,
    });

  const [debouncedFocusState] = useDebounce(focusState, 0);

  const doClearInputValue = React.useCallback(() => {
    if (!multiple || autoClearSearchValue) {
      setInputValue('', { action: SelectSearchAction.AutoClear });
    }
  }, [autoClearSearchValue, setInputValue, multiple]);

  const handleSelectEvent = React.useCallback(
    (action: SelectAction) => {
      /**
       * 这里和 onChange 里都要调用 doClearInputValue
       * 1. 如果点击一个已选中的项目，触发 onSelect 但是不触发 onChange，这时候是需要 doClearInputValue 的
       * 2. 如果调用 set 函数，但是没有给 action 的话，只触发 onChange 不触发 onSelect，这时候也是需要 doClearInputValue 的
       */
      doClearInputValue();
      // onSelect 回调
      if (action.type === SelectActionType.Select) {
        isFunction(onSelect) &&
          onSelect(action.payload as V, {
            option: getOption(action.payload as V),
          });
      } else if (action.type === SelectActionType.DeSelect) {
        isFunction(onDeselect) &&
          onDeselect(action.payload as V, {
            option: getOption(action.payload as V),
          });
      }
    },
    [doClearInputValue, getOption, onDeselect, onSelect],
  );

  const _onChange = useEventCallback((_options: V[], action: SelectAction) => {
    doClearInputValue();

    if (!isFunction(props.onChange)) {
      return;
    }

    if (props.mode === 'multiple') {
      const meta = {
        option: _options.map(opt => getOption(opt)),
        action,
      };
      props.onChange(_options, meta);
      return;
    }

    const option: OP | undefined = _options.map(opt => getOption(opt))[0];
    const meta = {
      option,
      action,
    };
    props.onChange(option?.value, meta);
  });

  const toggle = React.useCallback<SelectContextType<V, OP>['toggle']>(
    (_value, option = {}) => {
      if (!getPermittedOptionKeys([_value], disabled, optionMap).length) {
        return;
      }
      let action: SelectAction = {
        type: SelectActionType.Toggle,
        payload: undefined,
      };
      setValueSet(
        s => {
          if ('closeOverlay' in option) {
            if (option.closeOverlay) {
              setOverlayVisible(false);
            }
          } else if (!multiple) {
            setOverlayVisible(false);
          }
          setTimeout(() => {
            inputRef.current?.focus();
          });
          /**
           * Deselect
           * 1. single 不会 deselect
           * 2. multiple 如果存在 selectedMap 里，进行 deselect
           */
          if (valueSet.has(_value) && multiple) {
            action = {
              type: SelectActionType.DeSelect,
              payload: _value,
            };
            handleSelectEvent(action);
            return deselectReducer(_value, s).nextState;
          }
          /**
           * Select
           * 1. option list 里面有的话进行选中
           */
          if (optionMap.get(_value)) {
            action = {
              type: SelectActionType.Select,
              payload: _value,
            };
            handleSelectEvent(action);
            return selectReducer(optionMap.get(_value), multiple, s).nextState;
          }
          return s;
        },
        nextState => {
          _onChange([...nextState.keys()], action);
        },
      );
    },
    [_onChange, disabled, handleSelectEvent, inputRef, multiple, optionMap, setOverlayVisible, setValueSet, valueSet],
  );

  const select = toggle;
  const deselect = toggle;

  const set = React.useCallback<SelectContextType<V, OP>['set']>(
    (_value, option = {}) => {
      if (!multiple && _value.length > 1) {
        throw new Error('Selection in single mode can not set _value which length gt 1');
      }
      if ('closeOverlay' in option) {
        if (option.closeOverlay) {
          setOverlayVisible(false);
        }
      } else if (!multiple) {
        setOverlayVisible(false);
      }

      setTimeout(() => {
        inputRef.current?.focus();
      });

      // 单选不可以取消, 但需要触发 onSelect
      if (!multiple && option.action?.type === SelectActionType.DeSelect) {
        // 触发 onSelect
        option.action &&
          handleSelectEvent({
            ...option.action,
            type: SelectActionType.Select,
          });
        return;
      }

      setValueSet(
        () =>
          _value.reduce((pre, cur) => {
            const opt = getOption(cur);
            if (opt) {
              pre.add(cur);
            }
            return pre;
          }, new Set<V>()),
        nextState => {
          option.action && handleSelectEvent(option.action);
          _onChange(
            [...nextState.keys()],
            option.action ?? {
              type: SelectActionType.Set,
              payload: undefined,
            },
          );
        },
      );
    },
    [_onChange, getOption, handleSelectEvent, inputRef, multiple, setOverlayVisible, setValueSet],
  );

  const clear = React.useCallback<SelectContextType<V, OP>['clear']>(() => {
    setValueSet(
      () => new Set(),
      nextState => {
        setOverlayVisible(false);
        setTimeout(() => {
          inputRef.current?.focus();
        });
        onClear?.();
        _onChange([...nextState.keys()], {
          type: SelectActionType.Clear,
          payload: undefined,
        });
      },
    );
  }, [_onChange, inputRef, onClear, setOverlayVisible, setValueSet]);

  const deselectLast = React.useCallback<SelectContextType<V, OP>['deselectLast']>(() => {
    const keys = [...valueSet.keys()];
    const keyToDeselect = keys[keys.length - 1];
    if (keyToDeselect) {
      deselect(keyToDeselect);
    }
  }, [deselect, valueSet]);

  const {
    onInputChange,
    onInputCompositionEnd,
    onInputCompositionStart,
    onInputPaste,
    onInternalKeyUp,
    onInternalKeyDown,
    compositionStatusRef,
  } = useSelectInput({
    multiple,
    tokenSeparators,
    inputValue,
    setInputValue,
    set,
    deselectLast,
    overlayVisible,
    setOverlayVisible,
    onKeyDown,
    onKeyUp,
    listRef,
    setInputting,
    getOption,
  });

  const cssScope = useCssScope(`${config.selectorCssScope}`);
  const dropdownCssScope = useCssScope(`${config.dropdownCssScope}__dropdown`);
  const selectorCssScope = useCssScope(`${config.selectorCssScope}__selector`);

  // ================ props.optionLabelProp ================
  let { optionLabelProp } = props;
  // 这里是有可能出现 children 的
  if (optionLabelProp === 'children' || optionLabelProp === undefined) {
    optionLabelProp = config.defaultOptionLabelProp;
  }
  const selectorLabelRenderer = useCallback<SelectContextType<V, OP>['selectorLabelRenderer']>(
    _value => {
      const option = getOption(_value);
      if (isFunction(optionLabelProp)) {
        return optionLabelProp(option);
      }
      const node = (option as unknown as any)[optionLabelProp ?? 'label'];
      return node;
    },
    [getOption, optionLabelProp],
  );

  const tagRenderer = tagRenderProp ?? tagRendererProp ?? defaultTagRenderer;

  return {
    focusState,
    disabled,
    blur,
    focus,
    inputRef,
    listRef,
    inputValue,
    overlayVisible,
    multiple,
    valueSet,
    getOption,
    set,
    select,
    deselect,
    toggle,
    clear,
    deselectLast,
    setInputValue,
    setOverlayVisible,
    selectorCssScope,
    dropdownCssScope,
    onContainerFocus,
    onContainerBlur,
    onInputChange,
    onInputCompositionEnd,
    onInputCompositionStart,
    onInputPaste,
    onInternalKeyDown,
    onInternalKeyUp,
    compositionStatusRef,
    showSearch,
    tabIndex,
    selectorLabelRenderer,
    cssScope,
    size,
    cancelDelayCloseOverlay,
    tagRenderer,
    id,
    // 若使用 focusState 来控制 inputting，在点击 option 时会短暂失焦，导致输入框中的内容闪动。因此改为 debounced 后的 focusState
    // inputting: showSearch && focusState && inputting,
    inputting: showSearch && debouncedFocusState && inputting,
    setInputting,
    noOptionsContent,
    notFoundContent,
    dropdownRender,
    loading,
    textOverflow: dropdownMatchSelectWidth === true,
  };
}
